// @farris/ui-shortcuts/esm2015/lib/shortcuts.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { codes, modifiers } from './keys';
import { Subject } from 'rxjs';
import { filter } from 'rxjs/operators';
import { allPass, any, difference, identity, isNill } from './utils';
import * as i0 from "@angular/core";
/** @type {?} */
const $$ngOnDestroy = Symbol('OnDestroy');
export class ShortcutsService {
    constructor() {
        /**
         * 解析快捷键
         * 并为每个KEY 创建一个函数
         */
        this._shortcuts = [];
        /**
         * 控制按键时间.
         */
        this.throttleTime = 0;
        this._pressed = new Subject();
        /**
         * Streams of pressed events, can be used instead or with a command.
         */
        this.pressed$ = this._pressed.asObservable();
        /**
         * 禁用快捷键
         */
        this.disabled = false;
        this._ignored = ['INPUT', 'TEXTAREA', 'SELECT'];
        this.isAllowed = (/**
         * @param {?} shortcut
         * @return {?}
         */
        (shortcut) => {
            /** @type {?} */
            const target = (/** @type {?} */ (shortcut.event.target));
            if (target === shortcut.target) {
                return true;
            }
            if (shortcut.allowIn.length) {
                return !((/** @type {?} */ (difference(this._ignored, shortcut.allowIn)))).includes(target.nodeName);
            }
            return !((/** @type {?} */ (this._ignored))).includes(target.nodeName);
        });
        this.mapEvent = (/**
         * @param {?} event
         * @return {?}
         */
        event => this._shortcuts
            .map((/**
         * @param {?} shortcut
         * @return {?}
         */
        shortcut => Object.assign({}, shortcut, {
            predicates: any(identity, shortcut.predicates.map((/**
             * @param {?} predicates
             * @return {?}
             */
            (predicates) => allPass(predicates)(event)))),
            event: event
        })))
            .filter((/**
         * @param {?} shortcut
         * @return {?}
         */
        shortcut => shortcut.predicates))
            .reduce((/**
         * @param {?} acc
         * @param {?} shortcut
         * @return {?}
         */
        (acc, shortcut) => (acc.priority > shortcut.priority ? acc : shortcut)), (/** @type {?} */ ({
            priority: 0
        }))));
        /**
         * transforms a shortcut to:
         * a predicate function
         */
        this.getKeys = (/**
         * @param {?} command
         * @return {?}
         */
        (command) => command.map((/**
         * @param {?} key
         * @return {?}
         */
        key => key.trim())).filter((/**
         * @param {?} key
         * @return {?}
         */
        key => key !== '+'))
            .map((/**
         * @param {?} key
         * @return {?}
         */
        key => {
            // for modifiers like control key
            // look for event['ctrlKey']
            // otherwise use the keyCode
            if (modifiers.hasOwnProperty(key)) {
                return (/**
                 * @param {?} event
                 * @return {?}
                 */
                event => !!event[modifiers[key]]);
            }
            return (/**
             * @param {?} event
             * @return {?}
             */
            event => codes[key]
                ? event.keyCode === codes[key] || event.key === key
                : event.keyCode === key.toUpperCase().charCodeAt(0));
        })));
        // this.keydown$ = fromEvent(document, 'keydown').pipe(
        //     filter(_ => !this.disabled),
        //     map(this.mapEvent),
        //     filter(
        //         (shortcut: ParsedShortcut) =>
        //             !shortcut.target || shortcut.event.target === shortcut.target
        //     ),
        //     filter((shortcut: ParsedShortcut) => isFunction(shortcut.command)),
        //     filter(this.isAllowed),
        //     tap((shortcut: any) => !shortcut.preventDefault || shortcut.event.preventDefault()),
        //     debounce(shortcut => timer(shortcut.throttleTime)),
        //     tap(shortcut => shortcut.command({ event: shortcut.event, key: shortcut.key })),
        //     tap(shortcut => this._pressed.next({ event: shortcut.event, key: shortcut.key })),
        //     catchError(error => of(error))
        // );
        // this.subscription = this.keydown$.subscribe();
    }
    /**
     * @private
     * @return {?}
     */
    get shortcuts() {
        return this._shortcuts;
    }
    /**
     * Remove subscription.
     * @return {?}
     */
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
    /**
     * Enable all keyboard shortcuts
     * @return {?}
     */
    enable() {
        this.disabled = false;
        return this;
    }
    /**
     * Disable all keyboard shortcuts
     * @return {?}
     */
    disable() {
        this.disabled = true;
        return this;
    }
    /**
     * Check if all keyboard shortcuts are disabled.
     * @return {?}
     */
    isDisabled() {
        return this.disabled;
    }
    /**
     * Add new shortcut/s
     * @param {?} shortcuts
     * @param {?=} instance
     * @return {?}
     */
    add(shortcuts, instance) {
        shortcuts = Array.isArray(shortcuts) ? shortcuts : [shortcuts];
        if (instance) {
            const [key] = [...shortcuts.map((/**
                 * @param {?} shortcut
                 * @return {?}
                 */
                shortcut => shortcut.key))];
            this.bindOnDestroy(instance, key);
        }
        this._shortcuts.push(...this.parseCommand(shortcuts));
        return this;
    }
    /**
     * bind to the component ngOnDestroy to remove related keys
     * when component is destroyed.
     * @private
     * @param {?} instance - component to remove keys when ngOnDestroy is called.
     * @param {?} keys
     * @return {?}
     */
    bindOnDestroy(instance, keys) {
        if (instance.ngOnDestroy) {
            instance[$$ngOnDestroy] = instance.ngOnDestroy;
        }
        /** @type {?} */
        const that = this;
        instance.ngOnDestroy = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            const onDestroy = instance[$$ngOnDestroy];
            if (onDestroy) {
                onDestroy.apply(this);
            }
            that.remove(keys);
        });
        return this;
    }
    /**
     * Remove a command based on key or array of keys.
     * can be used for cleanup.
     * @param {?} key
     * @return {?}
     */
    remove(key) {
        /** @type {?} */
        const keys = Array.isArray(key) ? key : [key];
        this._shortcuts = this._shortcuts.filter((/**
         * @param {?} shortcut
         * @return {?}
         */
        shortcut => {
            return !shortcut.key.find((/**
             * @param {?} sKey
             * @return {?}
             */
            sKey => {
                return keys.filter((/**
                 * @param {?} k
                 * @return {?}
                 */
                k => k === sKey)).length > 0;
            }));
        }));
        return this;
    }
    /**
     * Returns an observable of keyboard shortcut filtered by a specific key.
     * @param {?} key - the key to filter the observable by.
     * @return {?}
     */
    select(key) {
        return this.pressed$.pipe(filter((/**
         * @param {?} __0
         * @return {?}
         */
        ({ event, key: eventKeys }) => {
            return !!eventKeys.find((/**
             * @param {?} eventKey
             * @return {?}
             */
            eventKey => eventKey === key));
        })));
    }
    /**
     * Parse each command using getKeys function
     * @private
     * @param {?} command
     * @return {?}
     */
    parseCommand(command) {
        /** @type {?} */
        const commands = Array.isArray(command) ? command : [command];
        return commands.map((/**
         * @param {?} cmd
         * @return {?}
         */
        (cmd) => {
            /** @type {?} */
            const keys = Array.isArray(cmd.key) ? cmd.key : [cmd.key];
            /** @type {?} */
            const priority = Math.max(...keys.map((/**
             * @param {?} key
             * @return {?}
             */
            key => key.split(' ').length)));
            /** @type {?} */
            const predicates = keys.map((/**
             * @param {?} key
             * @return {?}
             */
            key => this.getKeys(key.split(' '))));
            return (/** @type {?} */ (Object.assign({}, cmd, { allowIn: cmd.allowIn || [], key: keys, throttle: isNill(cmd.throttleTime) ? this.throttleTime : cmd.throttleTime, priority: priority, predicates: predicates })));
        }));
    }
}
ShortcutsService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ShortcutsService.ctorParameters = () => [];
/** @nocollapse */ ShortcutsService.ngInjectableDef = i0.defineInjectable({ factory: function ShortcutsService_Factory() { return new ShortcutsService(); }, token: ShortcutsService, providedIn: "root" });
if (false) {
    /**
     * 解析快捷键
     * 并为每个KEY 创建一个函数
     * @type {?}
     * @private
     */
    ShortcutsService.prototype._shortcuts;
    /**
     * 控制按键时间.
     * @type {?}
     * @private
     */
    ShortcutsService.prototype.throttleTime;
    /**
     * @type {?}
     * @private
     */
    ShortcutsService.prototype._pressed;
    /**
     * Streams of pressed events, can be used instead or with a command.
     * @type {?}
     */
    ShortcutsService.prototype.pressed$;
    /**
     * 禁用快捷键
     * @type {?}
     * @private
     */
    ShortcutsService.prototype.disabled;
    /**
     * @type {?}
     * @private
     */
    ShortcutsService.prototype._ignored;
    /**
     * Subscription for on destroy.
     * @type {?}
     * @private
     */
    ShortcutsService.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    ShortcutsService.prototype.keydown$;
    /**
     * @type {?}
     * @private
     */
    ShortcutsService.prototype.isAllowed;
    /**
     * @type {?}
     * @private
     */
    ShortcutsService.prototype.mapEvent;
    /**
     * transforms a shortcut to:
     * a predicate function
     * @type {?}
     * @private
     */
    ShortcutsService.prototype.getKeys;
}
