// @farris/ui-shortcuts/esm2015/lib/utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} x
 * @return {?}
 */
export function isFunction(x) {
    return typeof x === 'function';
}
/** @type {?} */
export const any = (/**
 * @param {?} fn
 * @param {?} list
 * @return {?}
 */
(fn, list) => {
    /** @type {?} */
    let idx = 0;
    while (idx < list.length) {
        if (fn(list[idx])) {
            return true;
        }
        idx += 1;
    }
    return false;
});
/** @type {?} */
export const identity = (/**
 * @param {?} x
 * @return {?}
 */
x => x);
/** @type {?} */
export const isNill = (/**
 * @param {?} x
 * @return {?}
 */
x => x == null);
/** @type {?} */
export const difference = (/**
 * @param {?} first
 * @param {?} second
 * @return {?}
 */
(first, second) => first.filter((/**
 * @param {?} item
 * @return {?}
 */
item => !((/** @type {?} */ (second))).some((/**
 * @param {?} secondItem
 * @return {?}
 */
secondItem => secondItem == item)))));
/** @type {?} */
export const allPass = (/**
 * @param {?} preds
 * @return {?}
 */
preds => (/**
 * @param {...?} args
 * @return {?}
 */
(...args) => {
    /** @type {?} */
    let idx = 0;
    /** @type {?} */
    const len = preds.length;
    while (idx < len) {
        if (!preds[idx].apply(this, args)) {
            return false;
        }
        idx += 1;
    }
    return true;
}));
