// @gsp-cmp/chgdr/esm2015/lib/service/chgdr-inner.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/chgdr-inner.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { HttpService, SessionService } from '@ecp-caf/caf-common';
import { GSPMetadataRTService } from '@gsp-lcm/metadatart-selector';
import { forkJoin, of } from 'rxjs';
import { map } from 'rxjs/operators';
import { ChgdrConfigHandler } from './chgdr-config-handler';
import { ChgdrConfigService } from './chgdr-config.service';
import { ChgdrMetadataService } from './chgdr-metadata.service';
import * as i0 from "@angular/core";
import * as i1 from "@ecp-caf/caf-common";
import * as i2 from "@gsp-lcm/metadatart-selector";
import * as i3 from "./chgdr-metadata.service";
import * as i4 from "./chgdr-config.service";
export class ChgdrInnerService {
    /**
     * @param {?} http
     * @param {?} sessionService
     * @param {?} metadataService
     * @param {?} chgdrMetadataService
     * @param {?} chgdrConfigService
     * @param {?} injector
     */
    constructor(http, sessionService, metadataService, chgdrMetadataService, chgdrConfigService, injector) {
        this.http = http;
        this.sessionService = sessionService;
        this.metadataService = metadataService;
        this.chgdrMetadataService = chgdrMetadataService;
        this.chgdrConfigService = chgdrConfigService;
        this.injector = injector;
        this.beMetadataCache = new Map();
    }
    /**
     * 获取BE元数据
     * @param {?} beId
     * @return {?}
     */
    getChgdrConfigHandler(beId) {
        if (this.beMetadataCache.has(beId)) {
            /** @type {?} */
            let be = this.beMetadataCache.get(beId);
            return of(be);
        }
        else {
            /** @type {?} */
            let requests = [];
            requests.push(this.chgdrMetadataService.getGSPBusinessEntity(beId));
            requests.push(this.chgdrConfigService.getBeChgdrConfig(beId));
            return forkJoin(requests).pipe(map((/**
             * @param {?} data
             * @return {?}
             */
            data => {
                /** @type {?} */
                const be = data[0];
                /** @type {?} */
                const entityConfigs = data[1] || [];
                /** @type {?} */
                const handler = new ChgdrConfigHandler(be, entityConfigs);
                this.beMetadataCache.set(beId, handler);
                return handler;
            })));
        }
    }
}
ChgdrInnerService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
ChgdrInnerService.ctorParameters = () => [
    { type: HttpService },
    { type: SessionService },
    { type: GSPMetadataRTService },
    { type: ChgdrMetadataService },
    { type: ChgdrConfigService },
    { type: Injector }
];
/** @nocollapse */ ChgdrInnerService.ngInjectableDef = i0.defineInjectable({ factory: function ChgdrInnerService_Factory() { return new ChgdrInnerService(i0.inject(i1.HttpService), i0.inject(i1.SessionService), i0.inject(i2.GSPMetadataRTService), i0.inject(i3.ChgdrMetadataService), i0.inject(i4.ChgdrConfigService), i0.inject(i0.INJECTOR)); }, token: ChgdrInnerService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.beMetadataCache;
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.sessionService;
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.metadataService;
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.chgdrMetadataService;
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.chgdrConfigService;
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.injector;
}
