// @gsp-cmp/chgdr/esm2015/lib/service/chgdr-list-ui.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/chgdr-list-ui.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ComponentFactoryResolver, Injector } from '@angular/core';
import { MessagerService } from '@farris/ui-messager';
import { BsModalService } from '@farris/ui-modal';
import { of } from 'rxjs';
import { ChgdrListDialogContentComponent } from '../chgdr-list-dialog-content/chgdr-list-dialog-content.component';
import { ChgdrListDialogConfig } from '../model/chgdr-ui-config';
import { I18nUtil } from './i18n-service';
import * as i0 from "@angular/core";
import * as i1 from "@farris/ui-modal";
import * as i2 from "@farris/ui-messager";
export class ChgdrListUiService {
    /**
     * @param {?} resolver
     * @param {?} injector
     * @param {?} modalService
     * @param {?} messagerService
     */
    constructor(resolver, injector, modalService, messagerService) {
        this.resolver = resolver;
        this.injector = injector;
        this.modalService = modalService;
        this.messagerService = messagerService;
    }
    /**
     * @param {?} arg1
     * @param {?=} dataId
     * @return {?}
     */
    showDialog(arg1, dataId) {
        if (!arg1) {
            console.error("创建业务变更日志列表组件失败，业务实体不能为空，请检查", arg1, dataId);
            throw new Error("创建业务变更日志列表组件失败，业务实体不能为空，请检查");
        }
        if (arg1 instanceof String || typeof arg1 == "string") {
            /** @type {?} */
            let config = new ChgdrListDialogConfig();
            config.entityId = arg1.toString();
            config.dataId = dataId;
            this.create(config);
        }
        else {
            this.create(arg1);
        }
    }
    /**
     * @private
     * @param {?} config
     * @return {?}
     */
    create(config) {
        if (config.entityId == null) {
            console.error("创建业务变更日志列表组件失败，业务实体不能为空，请检查", config);
            throw new Error("创建业务变更日志列表组件失败，业务实体不能为空，请检查");
        }
        /** @type {?} */
        let fac = this.resolver.resolveComponentFactory(ChgdrListDialogContentComponent);
        /** @type {?} */
        let cmpRef = fac.create(this.injector);
        /** @type {?} */
        let modalConfig = {
            title: I18nUtil.instant("chgdr-list-dialog-title"),
            width: 1000,
            height: 550,
            resizable: true,
            showButtons: false,
            showMaxButton: true,
            beforeOpen: (/**
             * @return {?}
             */
            () => {
                return of(true);
            })
        };
        cmpRef.instance.entityId = config.entityId;
        cmpRef.instance.dataId = config.dataId;
        cmpRef.instance.beforeShowChangeDetail = config.beforeShowChangeDetail;
        /** @type {?} */
        let dialog = this.modalService.show(cmpRef, modalConfig);
    }
}
ChgdrListUiService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
ChgdrListUiService.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: Injector },
    { type: BsModalService },
    { type: MessagerService }
];
/** @nocollapse */ ChgdrListUiService.ngInjectableDef = i0.defineInjectable({ factory: function ChgdrListUiService_Factory() { return new ChgdrListUiService(i0.inject(i0.ComponentFactoryResolver), i0.inject(i0.INJECTOR), i0.inject(i1.BsModalService), i0.inject(i2.MessagerService)); }, token: ChgdrListUiService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ChgdrListUiService.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    ChgdrListUiService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    ChgdrListUiService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    ChgdrListUiService.prototype.messagerService;
}
