// @gsp-cmp/chgdr/esm2015/lib/service/i18n-service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/i18n-service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { TranslateDefaultParser } from '@ngx-translate/core';
import { LangDataEn } from './i18n/en';
import { LangDataZhChs } from './i18n/zh-CHS';
import { LangDataZhCht } from './i18n/zh-CHT';
import * as i0 from "@angular/core";
/** @type {?} */
let I18nServiceInstance;
/** @type {?} */
let langResourceMap = new Map();
langResourceMap.set("en", LangDataEn);
langResourceMap.set("zh-CHS", LangDataZhChs);
langResourceMap.set("zh-CHT", LangDataZhCht);
/** @type {?} */
let currentLang = localStorage.getItem('languageCode');
if (!currentLang) {
    currentLang = "zh-CHS";
    console.log("未找到设置的语言，默认使用" + currentLang);
}
/** @type {?} */
let parser = new TranslateDefaultParser();
export class I18nUtil {
    /**
     * @param {?} key
     * @param {?=} param
     * @return {?}
     */
    static instant(key, param) {
        /** @type {?} */
        let statement = I18nUtil.getCurrentLangResource()[key];
        if (statement === undefined || statement === null) {
            return key;
        }
        /** @type {?} */
        let result = parser.interpolate(statement, param);
        if (result === undefined || result === null) {
            return key;
        }
        return result;
    }
    /**
     * @private
     * @return {?}
     */
    static getCurrentLangResource() {
        return langResourceMap.get(currentLang) || {};
    }
    /**
     * @param {?} key
     * @param {?} defaultText
     * @return {?}
     */
    static instantOrDefault(key, defaultText) {
        /** @type {?} */
        let statement = I18nUtil.getCurrentLangResource()[key];
        if (statement === undefined || statement === null) {
            return defaultText;
        }
        /** @type {?} */
        let result = parser.interpolate(statement);
        if (result === undefined || result === null) {
            return defaultText;
        }
        return result;
    }
}
export class I18nService {
    constructor() {
    }
    /**
     * @param {?} key
     * @param {?=} param
     * @return {?}
     */
    instant(key, param) {
        return I18nUtil.instant(key, param);
    }
    /**
     * @private
     * @return {?}
     */
    getCurrentLangResource() {
        return langResourceMap.get(currentLang) || {};
    }
    /**
     * @param {?} key
     * @param {?} defaultText
     * @return {?}
     */
    static staticInstant(key, defaultText) {
        return I18nUtil.instantOrDefault(key, defaultText);
    }
}
I18nService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
I18nService.ctorParameters = () => [];
/** @nocollapse */ I18nService.ngInjectableDef = i0.defineInjectable({ factory: function I18nService_Factory() { return new I18nService(); }, token: I18nService, providedIn: "root" });
