// @gsp-wf/wf-sign/esm5/lib/sign/sign.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/sign/sign.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, ElementRef, Input, Output, EventEmitter } from '@angular/core';
import { SignatureComponent } from '../signature/signature.component';
var SignComponent = /** @class */ (function () {
    function SignComponent() {
        // 抛出选中的事件
        this.toDataUrlEvent = new EventEmitter();
        this.isDrawEvent = new EventEmitter();
        this.imageType = "image/png";
        this.imageQuality = 1;
        this.isDraw = false;
        this.signaturePadOptions = {
            // passed through to szimek/signature_pad constructor
            'minWidth': 1,
            'canvasWidth': 500,
            'canvasHeight': 240
        };
    }
    /**
     * @return {?}
     */
    SignComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    SignComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        // this.signaturePad is now available
        this.signaturePad.set('minWidth', 1); // set szimek/signature_pad options at runtime
        this.signaturePad.set('canvasWidth', this.canvasDiv.nativeElement.clientWidth);
        this.signaturePad.clear(); // invoke functions from szimek/signature_pad API
        this.signaturePad.setImageBackground();
    };
    /**
     * @return {?}
     */
    SignComponent.prototype.confirm = /**
     * @return {?}
     */
    function () {
        console.log(this.signaturePad.toDataURL(this.imageType, this.imageQuality));
        this.signaturePad.setWaterMark(new Date());
        this.toDataUrlEvent.emit(this.signaturePad.toDataURL(this.imageType, this.imageQuality));
    };
    /**
     * @return {?}
     */
    SignComponent.prototype.reset = /**
     * @return {?}
     */
    function () {
        this.signaturePad.clear();
        this.isDraw = false;
        this.isDrawEvent.emit(this.isDraw);
    };
    /**
     * @return {?}
     */
    SignComponent.prototype.drawComplete = /**
     * @return {?}
     */
    function () {
        this.isDraw = true;
        this.isDrawEvent.emit(this.isDraw);
        // will be notified of szimek/signature_pad's onEnd event
        //console.log(this.signaturePad.toDataURL());
    };
    /**
     * @return {?}
     */
    SignComponent.prototype.drawStart = /**
     * @return {?}
     */
    function () {
        // will be notified of szimek/signature_pad's onBegin event
        console.log('begin drawing');
    };
    SignComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-sign-sign',
                    template: "<div class=\"fm-feature-suggestion fm-feature-section--gutter\">\r\n    <div #canvastitle style=\"padding:2px 2px 2px 0;text-align:left;width:100%;height: 21px;\">\r\n        <span style=\"font-size: 15px;color: #999999;\">\u7B7E\u540D\u533A\u57DF</span>\r\n      </div>\r\n      <wf-sign-signature  [options]=\"signaturePadOptions\" (onBeginEvent)=\"drawStart()\" (onEndEvent)=\"drawComplete()\"></wf-sign-signature>\r\n  <div class=\"fm-feature-suggestion--bar\">\r\n    <ul class=\"suggestion--bar-common\">\r\n      <li class=\"suggestion--bar-item\" (click)=\"confirm()\">\r\n        <span>\u786E\u8BA4</span>\r\n      </li>\r\n      <li class=\"suggestion--bar-item\" (click)=\"reset()\">\r\n        <span>\u91CD\u7F6E</span>\r\n      </li>\r\n    </ul>\r\n  </div>\r\n</div>\r\n",
                    styles: [".fm-page-main1{overflow:hidden}.fm-utils-flex-column{display:flex;flex-direction:column}.fm-page{position:absolute;top:0;left:0;bottom:0;right:0;display:flex;flex-direction:column;background:#fff;font-size:13px;line-height:18px}.fm-page-main{flex-basis:0;flex-shrink:1;flex-grow:1;overflow-x:hidden;overflow-y:auto;position:relative}.fm-feature-section--gutter{margin:0 0 12px}.fm-feature-suggestion{background:#fff;padding:12px 16px}.fm-feature-suggestion .fm-feature-suggestion--bar{display:flex;flex-direction:row;align-items:center;color:#888}"]
                }] }
    ];
    /** @nocollapse */
    SignComponent.ctorParameters = function () { return []; };
    SignComponent.propDecorators = {
        signaturePad: [{ type: ViewChild, args: [SignatureComponent,] }],
        canvasDiv: [{ type: ViewChild, args: ['canvastitle',] }],
        toDataUrlEvent: [{ type: Output }],
        isDrawEvent: [{ type: Output }],
        imageType: [{ type: Input }],
        imageQuality: [{ type: Input }]
    };
    return SignComponent;
}());
export { SignComponent };
if (false) {
    /** @type {?} */
    SignComponent.prototype.signaturePad;
    /** @type {?} */
    SignComponent.prototype.canvasDiv;
    /** @type {?} */
    SignComponent.prototype.toDataUrlEvent;
    /** @type {?} */
    SignComponent.prototype.isDrawEvent;
    /** @type {?} */
    SignComponent.prototype.imageType;
    /** @type {?} */
    SignComponent.prototype.imageQuality;
    /**
     * @type {?}
     * @private
     */
    SignComponent.prototype.isDraw;
    /** @type {?} */
    SignComponent.prototype.signaturePadOptions;
}
