// @farris/devkit/esm5/lib/command/command_bus.js
/**
 * CommandBus相关定义
 * @author Witt<jiwt@inspur.com>
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject, Subject } from 'rxjs';
import { CommandHandlerFactory } from './command_handler_factory';
/**
 * CommandBus用于派发Command，它接受一个Command实例，查找对应的CommandHandler，并执行。
 */
var CommandBus = /** @class */ (function () {
    /**
     * 构造函数
     */
    function CommandBus(handlerFactory) {
        this.handlerFactory = handlerFactory;
        this.executingCommands = [];
        this.executingCommandCount$ = new BehaviorSubject(this.executingCommands.length);
    }
    /**
     * 派发命令
     * @param command 要派发的命令
     */
    CommandBus.prototype.dispatch = function (command) {
        var _this = this;
        var commandResult$ = new Subject();
        this.executeCommand(command).subscribe({
            next: function (lastTaskResult) {
                commandResult$.next(lastTaskResult);
                commandResult$.complete();
            },
            complete: function () {
                commandResult$.complete();
                _this.removeCommandFromExecutingQueue(command);
            },
            error: function (error) {
                commandResult$.error(error);
                _this.removeCommandFromExecutingQueue(command, !_this.is401Error(error));
            }
        });
        return commandResult$;
    };
    /**
     * dispose
     */
    CommandBus.prototype.dispose = function () {
        this.handlerFactory.dispose();
    };
    /**
     * 执行命令并返回最后一个任务的执行结果流
     */
    CommandBus.prototype.executeCommand = function (command) {
        this.addCommandToExecutingQueue(command);
        var commandName = command.name;
        var handler = this.handlerFactory.create(commandName);
        var lastTaskResult$ = handler.execute(command);
        return lastTaskResult$;
    };
    /**
     * 添加到执行队列
     */
    CommandBus.prototype.addCommandToExecutingQueue = function (command) {
        this.executingCommands.push(command);
        this.executingCommandCount$.next(this.executingCommands.length);
    };
    /**
     * 从执行队列中移除
     */
    CommandBus.prototype.removeCommandFromExecutingQueue = function (command, postOrNot) {
        if (postOrNot === void 0) { postOrNot = true; }
        this.executingCommands = this.executingCommands.filter(function (executingCommand) {
            return executingCommand !== command;
        });
        if (postOrNot === true) {
            this.executingCommandCount$.next(this.executingCommands.length);
        }
    };
    CommandBus.prototype.is401Error = function (error) {
        // 如果error为空，不认为是401错误
        return error && error.status === 401;
    };
    CommandBus.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CommandBus.ctorParameters = function () { return [
        { type: CommandHandlerFactory }
    ]; };
    return CommandBus;
}());
export { CommandBus };
