// @farris/ui-dropdown/esm5/lib/component/dropdown.item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
var FarrisDropdownItemComponent = /** @class */ (function () {
    function FarrisDropdownItemComponent() {
        this.select = new EventEmitter();
    }
    /**
     * @return {?}
     */
    FarrisDropdownItemComponent.prototype.selectItem = /**
     * @return {?}
     */
    function () {
        if (this.model.disabled) {
            return;
        }
        this.select.emit(this.model);
    };
    FarrisDropdownItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-dropdown-item',
                    template: "\n        <div class=\"dropdown-divider\" *ngIf=\"model.divide\"></div>\n        <a\n            class=\"dropdown-item\"\n            [class.active]=\"model.active\"\n            [class.disabled]=\"model.disabled\"\n            href=\"javascript:void(0)\"\n            (click)=\"selectItem()\"\n            >{{ model.label }}</a\n        >\n    "
                }] }
    ];
    FarrisDropdownItemComponent.propDecorators = {
        model: [{ type: Input }],
        select: [{ type: Output }]
    };
    return FarrisDropdownItemComponent;
}());
export { FarrisDropdownItemComponent };
if (false) {
    /** @type {?} */
    FarrisDropdownItemComponent.prototype.model;
    /** @type {?} */
    FarrisDropdownItemComponent.prototype.select;
}
