// @farris/kendo-binding/esm5/lib/directives/farris-list-filter/farris_list_filter_binding.directive.js
import { Directive, Injector, Input, Optional } from '@angular/core';
import { FrameContext } from '@farris/devkit';
import { ListFilterComponent } from '@farris/ui-list-filter';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
var FarrisListFilterBindingDirective = /** @class */ (function () {
    function FarrisListFilterBindingDirective(injector, frameContext, listFilterComponent) {
        this.injector = injector;
        this.frameContext = frameContext;
        this.listFilterComponent = listFilterComponent;
        /**
         * 变量编号，默认为filterConditionList
         */
        this.variableCode = 'filterConditionList';
        this.$destroy = new Subject();
    }
    FarrisListFilterBindingDirective.prototype.ngOnDestroy = function () {
        if (this.$destroy) {
            this.$destroy.next();
            this.$destroy.complete();
        }
    };
    FarrisListFilterBindingDirective.prototype.ngOnInit = function () {
        var _this = this;
        if (this.listFilterComponent) {
            this.listFilterComponent.conditionsChange.pipe(takeUntil(this.$destroy)).subscribe(function (event) {
                var _a = (event || {}).data, data = _a === void 0 ? '[]' : _a;
                _this.setConditions(data);
            });
        }
    };
    FarrisListFilterBindingDirective.prototype.setConditions = function (conditions) {
        this.frameContext.uiState.setPropertyValue(this.variableCode, conditions);
    };
    FarrisListFilterBindingDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-list-filter-binding]'
                },] }
    ];
    /** @nocollapse */
    FarrisListFilterBindingDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: FrameContext },
        { type: ListFilterComponent, decorators: [{ type: Optional }] }
    ]; };
    FarrisListFilterBindingDirective.propDecorators = {
        variableCode: [{ type: Input, args: ['variableCode',] }]
    };
    return FarrisListFilterBindingDirective;
}());
export { FarrisListFilterBindingDirective };
