// @gsp-svc/filtercondition/esm2015/lib/NewSchema.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * \@description 统一结构Schema
 * @param name 名称
 * @param description 描述
 * @param entitySets 实体集
 * @param entityTypes 实体类型
 * @param associations 关联集合
 * \@date 2021-04-14
 * @author Wenxing Man
 */
export class NewSchema {
    constructor() {
        this.entitySets = [];
        this.entityTypes = [];
        this.associations = [];
    }
}
if (false) {
    /** @type {?} */
    NewSchema.prototype.name;
    /** @type {?} */
    NewSchema.prototype.description;
    /** @type {?} */
    NewSchema.prototype.entitySets;
    /** @type {?} */
    NewSchema.prototype.entityTypes;
    /** @type {?} */
    NewSchema.prototype.associations;
}
/**
 * @param code 实体集名
 * @param name 实体集语义化名
 * @param entityType 对应EntityType的id
 * \@date 2021-04-14
 * @author Wenxing Man
 */
export class NewEntitySet {
}
if (false) {
    /** @type {?} */
    NewEntitySet.prototype.code;
    /** @type {?} */
    NewEntitySet.prototype.name;
    /** @type {?} */
    NewEntitySet.prototype.entityType;
}
/**
 * @param code 属性名
 * @param name 实体语义化名
 * @param master  是否是主实体
 * @param properties    属性集合
 * @param keys  主键集合
 * @param navigationPorperties 导航集合属性
 * \@date 2021-04-14
 * @author Wenxing Man
 */
export class NewEntityType {
    constructor() {
        this.properties = [];
        this.keys = [];
        this.navigationProperties = [];
    }
}
if (false) {
    /** @type {?} */
    NewEntityType.prototype.code;
    /** @type {?} */
    NewEntityType.prototype.name;
    /** @type {?} */
    NewEntityType.prototype.master;
    /** @type {?} */
    NewEntityType.prototype.properties;
    /** @type {?} */
    NewEntityType.prototype.keys;
    /** @type {?} */
    NewEntityType.prototype.navigationProperties;
}
/**
 * @param name 关联名称
 * @param masterType 主对象类型
 * @param childType 子对象类型
 * @param masterProperty 主对象字段
 * @param childProperty 从对象字段
 * \@date 2021-04-14
 * @author Wenxing Man
 */
export class NewAssociation {
}
if (false) {
    /** @type {?} */
    NewAssociation.prototype.name;
    /** @type {?} */
    NewAssociation.prototype.masterType;
    /** @type {?} */
    NewAssociation.prototype.childType;
    /** @type {?} */
    NewAssociation.prototype.masterProperty;
    /** @type {?} */
    NewAssociation.prototype.childProperty;
}
/**
 * @param code 属性名
 * @param name 属性语义化名
 * @param dataType 属性值类型
 * @param nullable 是否允许为空值
 * @param dafaultValue 属性默认值
 * \@date 2021-04-14
 * @author Wenxing Man
 */
export class NewProperty {
}
if (false) {
    /** @type {?} */
    NewProperty.prototype.code;
    /** @type {?} */
    NewProperty.prototype.name;
    /** @type {?} */
    NewProperty.prototype.dataType;
    /** @type {?} */
    NewProperty.prototype.nullable;
    /** @type {?} */
    NewProperty.prototype.defaultValue;
}
/**
 * @param code 导航属性名称
 * @param name 导航属性语义化名
 * @param navigationName 导航对象的名称
 * @param navigationType 属性类型
 * @param relationShip 关联的对象
 * \@date 2021-04-14
 * @author Wenxing Man
 */
export class NewNavigationProperty {
}
if (false) {
    /** @type {?} */
    NewNavigationProperty.prototype.code;
    /** @type {?} */
    NewNavigationProperty.prototype.name;
    /** @type {?} */
    NewNavigationProperty.prototype.navigationName;
    /** @type {?} */
    NewNavigationProperty.prototype.navigationType;
    /** @type {?} */
    NewNavigationProperty.prototype.relationShip;
}
/** @enum {number} */
const NewDataType = {
    String: 0,
    Byte: 1,
    Boolean: 2,
    Date: 3,
    DateTime: 4,
    Decimal: 5,
    Integer: 6,
};
export { NewDataType };
NewDataType[NewDataType.String] = 'String';
NewDataType[NewDataType.Byte] = 'Byte';
NewDataType[NewDataType.Boolean] = 'Boolean';
NewDataType[NewDataType.Date] = 'Date';
NewDataType[NewDataType.DateTime] = 'DateTime';
NewDataType[NewDataType.Decimal] = 'Decimal';
NewDataType[NewDataType.Integer] = 'Integer';
/** @enum {string} */
const NewNavigationType = {
    EntitySet: "EntitySet",
    EntityType: "EntityType",
};
export { NewNavigationType };
