// @gsp-svc/filtercondition/esm2015/lib/schema.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * schema实体类
 * \@property entitySets  实体集集合
 * \@property entityTypes 实体类型集合
 */
export class FilterEntitySchema {
    constructor() {
        this.entitySets = [];
        this.entityTypes = [];
    }
}
if (false) {
    /** @type {?} */
    FilterEntitySchema.prototype.entitySets;
    /** @type {?} */
    FilterEntitySchema.prototype.entityTypes;
}
/**
 * 过滤实体类型
 * 用于多级嵌套实体显示
 * \@property name 实体名(用于实体间导航使用)
 * \@property displayName 展示名称
 * \@property properties 属性集
 * \@property navigationProperties 导航属性集
 */
export class FilterEntityType {
    constructor() {
        this.properties = [];
        this.navigationProperties = [];
    }
}
if (false) {
    /** @type {?} */
    FilterEntityType.prototype.name;
    /** @type {?} */
    FilterEntityType.prototype.displayName;
    /** @type {?} */
    FilterEntityType.prototype.properties;
    /** @type {?} */
    FilterEntityType.prototype.navigationProperties;
}
/**
 * 过滤实体类型集
 * 用于多级嵌套实体显示
 * \@property name 实体集名
 * \@property entityTypeName 包含的实体类型名
 */
export class FilterEntitySet {
}
if (false) {
    /** @type {?} */
    FilterEntitySet.prototype.name;
    /** @type {?} */
    FilterEntitySet.prototype.displayName;
    /** @type {?} */
    FilterEntitySet.prototype.entityTypeName;
}
