// @gsp-svc/filtercondition/esm2015/lib/typeinfo.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const ExpressTypeEnum = {
    Value: 0,
    Expression: 1,
};
export { ExpressTypeEnum };
ExpressTypeEnum[ExpressTypeEnum.Value] = 'Value';
ExpressTypeEnum[ExpressTypeEnum.Expression] = 'Expression';
/** @enum {number} */
const RelationTypeEnum = {
    And: 0,
    Or: 1,
    AddProperty: 2,
    AddRelation: 3,
};
export { RelationTypeEnum };
RelationTypeEnum[RelationTypeEnum.And] = 'And';
RelationTypeEnum[RelationTypeEnum.Or] = 'Or';
RelationTypeEnum[RelationTypeEnum.AddProperty] = 'AddProperty';
RelationTypeEnum[RelationTypeEnum.AddRelation] = 'AddRelation';
/** @enum {number} */
const CompareTypeEnum = {
    GreaterThan: 0,
    GreaterThanOrEqual: 1,
    LessThan: 2,
    LessThanOrEqual: 3,
    Equal: 4,
    NotEqual: 5,
    Contains: 6,
    NotContains: 7,
    StartsWith: 8,
    NotStartsWith: 9,
    EndsWith: 10,
    NotEndsWith: 11,
    IsNull: 12,
    IsNotNull: 13,
    In: 14,
    NotIn: 15,
};
export { CompareTypeEnum };
CompareTypeEnum[CompareTypeEnum.GreaterThan] = 'GreaterThan';
CompareTypeEnum[CompareTypeEnum.GreaterThanOrEqual] = 'GreaterThanOrEqual';
CompareTypeEnum[CompareTypeEnum.LessThan] = 'LessThan';
CompareTypeEnum[CompareTypeEnum.LessThanOrEqual] = 'LessThanOrEqual';
CompareTypeEnum[CompareTypeEnum.Equal] = 'Equal';
CompareTypeEnum[CompareTypeEnum.NotEqual] = 'NotEqual';
CompareTypeEnum[CompareTypeEnum.Contains] = 'Contains';
CompareTypeEnum[CompareTypeEnum.NotContains] = 'NotContains';
CompareTypeEnum[CompareTypeEnum.StartsWith] = 'StartsWith';
CompareTypeEnum[CompareTypeEnum.NotStartsWith] = 'NotStartsWith';
CompareTypeEnum[CompareTypeEnum.EndsWith] = 'EndsWith';
CompareTypeEnum[CompareTypeEnum.NotEndsWith] = 'NotEndsWith';
CompareTypeEnum[CompareTypeEnum.IsNull] = 'IsNull';
CompareTypeEnum[CompareTypeEnum.IsNotNull] = 'IsNotNull';
CompareTypeEnum[CompareTypeEnum.In] = 'In';
CompareTypeEnum[CompareTypeEnum.NotIn] = 'NotIn';
/**
 * 关系类型
 * 包含并且、或者、添加属性、添加关系
 * @type {?}
 */
export const RelationTypes = [
    { id: "并且" },
    { id: "或者" },
];
// export const CompareTypes: Array<{ text: string, value: CompareTypeEnum }> =
//   [
//     { text: "大于", value: CompareTypeEnum.GreaterThan },
//     { text: "大于等于", value: CompareTypeEnum.GreaterThanOrEqual },
//     { text: "小于", value: CompareTypeEnum.LessThan },
//     { text: "小于等于", value: CompareTypeEnum.LessThanOrEqual },
//     { text: "等于", value: CompareTypeEnum.Equal },
//     { text: "不等于", value: CompareTypeEnum.NotEqual },
//     { text: "包含", value: CompareTypeEnum.Contains },
//     { text: "开头是", value: CompareTypeEnum.StartsWith },
//     { text: "结尾是", value: CompareTypeEnum.EndsWith },
//     { text: "不包含", value: CompareTypeEnum.NotContains },
//     { text: "不包含开头是", value: CompareTypeEnum.NotStartsWith },
//     { text: "不包含结尾是", value: CompareTypeEnum.NotEndsWith },
//     { text: "为(空)", value: CompareTypeEnum.IsNull },
//     { text: "不为(空)", value: CompareTypeEnum.IsNotNull },
//     { text: "在", value: CompareTypeEnum.In },
//     { text: "不在", value: CompareTypeEnum.NotIn }
//   ];
/**
 * 比较类型
 * 包含常规的关系比较符
 * @type {?}
 */
export const CompareTypes = [
    { id: '大于' },
    { id: '大于等于' },
    { id: '小于' },
    { id: '小于等于' },
    { id: '等于' },
    { id: '不等于' },
    { id: '包含' },
    { id: '开头是' },
    { id: '结尾是' },
    { id: '不包含' },
    { id: '不包含开头是' },
    { id: '不包含结尾是' },
    { id: '为空' },
    { id: '不为(空)' },
    { id: '在' },
    { id: '不在' }
]
/**
 * 表达式类型
 * 包含值、表达式两种类型
 */
;
/**
 * 表达式类型
 * 包含值、表达式两种类型
 * @type {?}
 */
export const ExpressTypes = [
    { id: '值' },
    { id: '表达式' }
];
