// @qdp/ide-cmp/esm5/lib/service/print-integration/print-integration.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { PrintIntegrationManagerService } from '@qdp/common';
var PrintIntegrationService = /** @class */ (function () {
    function PrintIntegrationService(printIntegrationManager) {
        var _this = this;
        this.printIntegrationManager = printIntegrationManager;
        this.printIntegrationInfo = new Subject();
        this.printIntegrationManager.printIntegrationInfoChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.printIntegrationInfo.next(value);
        }));
    }
    /**
     * @param {?} schemaId
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    PrintIntegrationService.prototype.getPrintIntegration = /**
     * @param {?} schemaId
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (schemaId, queryId, queryRelativeUrl) {
        return this.printIntegrationManager.getPrintIntegration(schemaId, queryId, queryRelativeUrl);
    };
    /**
     * @param {?} printIntegrationInfo
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    PrintIntegrationService.prototype.savePrintIntegration = /**
     * @param {?} printIntegrationInfo
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (printIntegrationInfo, queryRelativeUrl) {
        return this.printIntegrationManager.savePrintIntegration(printIntegrationInfo, queryRelativeUrl);
    };
    /**
     * @param {?} schemaId
     * @param {?} queryID
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    PrintIntegrationService.prototype.deletPrintIntegration = /**
     * @param {?} schemaId
     * @param {?} queryID
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (schemaId, queryID, queryRelativeUrl) {
        return this.printIntegrationManager.deletePrintIntegration(schemaId, queryID, queryRelativeUrl);
    };
    PrintIntegrationService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    PrintIntegrationService.ctorParameters = function () { return [
        { type: PrintIntegrationManagerService }
    ]; };
    return PrintIntegrationService;
}());
export { PrintIntegrationService };
if (false) {
    /** @type {?} */
    PrintIntegrationService.prototype.printIntegrationInfo;
    /**
     * @type {?}
     * @private
     */
    PrintIntegrationService.prototype.printIntegrationManager;
}
