// @gspwidget/widget-devkit/esm5/types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var PropertyTypes = {
    Text: 'textOption',
    Bool: 'boolOption',
    Number: 'numberOption',
    Enum: 'enumOption',
    Color: 'colorOption',
    // Date = 'dateOption',
    Object: 'objectOption',
    Custom: 'customOption',
    SelectHelp: 'selectHelpOption',
    /** Help 得到的值的结构，取决于配置 HelpOption 后，最终得到的数据结构 */
    Help: 'helpOption',
    /** DataField 得到的值的值的类型为 DataField */
    DataField: 'dataFieldOption',
    Filter: 'filterOption',
    Html: 'htmlOption',
    DataParam: 'dataParamOption',
};
export { PropertyTypes };
/**
 * @record
 */
export function DataField() { }
if (false) {
    /** @type {?} */
    DataField.prototype.name;
    /** @type {?} */
    DataField.prototype.label;
}
/**
 * type，类型，默认 PropertyTypes.String
 * isArray，是否为数组，默认 false
 * 用 prop[prop.type] 可获取到 xxxOption
 * @record
 */
export function PropertyOption() { }
if (false) {
    /** @type {?|undefined} */
    PropertyOption.prototype.name;
    /** @type {?|undefined} */
    PropertyOption.prototype.displayName;
    /** @type {?|undefined} */
    PropertyOption.prototype.type;
    /** @type {?|undefined} */
    PropertyOption.prototype.isArray;
    /** @type {?|undefined} */
    PropertyOption.prototype.category;
    /**
     * 分组路径，可用 / 分割多层
     * @type {?|undefined}
     */
    PropertyOption.prototype.tip;
    /**
     * 定义如何根据其他 prop 的值，控制当前 item 是否显示。
     * 注意：curPropValues 是当前 property 所在的直接父对象内的（**即与自身平级的**）所有 propValues，不包含爸爸的同辈。
     * 仅应当根据同辈的值控制显示与否，所以某儿子不应该根据爸爸的某同辈而变化。
     * @type {?|undefined}
     */
    PropertyOption.prototype.ifShow;
    /** @type {?|undefined} */
    PropertyOption.prototype.textOption;
    /** @type {?|undefined} */
    PropertyOption.prototype.boolOption;
    /** @type {?|undefined} */
    PropertyOption.prototype.numberOption;
    /** @type {?|undefined} */
    PropertyOption.prototype.enumOption;
    /** @type {?|undefined} */
    PropertyOption.prototype.colorOption;
    /** @type {?|undefined} */
    PropertyOption.prototype.dateOption;
    /** @type {?|undefined} */
    PropertyOption.prototype.objectOption;
    /** @type {?|undefined} */
    PropertyOption.prototype.selectHelpOption;
    /** @type {?|undefined} */
    PropertyOption.prototype.helpOption;
    /** @type {?|undefined} */
    PropertyOption.prototype.dataFieldOption;
}
/** @enum {number} */
var HelpStructure = {
    table: 0,
    treeTable: 1,
    transfer: 2,
    treeTransfer: 3,
};
export { HelpStructure };
HelpStructure[HelpStructure.table] = 'table';
HelpStructure[HelpStructure.treeTable] = 'treeTable';
HelpStructure[HelpStructure.transfer] = 'transfer';
HelpStructure[HelpStructure.treeTransfer] = 'treeTransfer';
/**
 * @record
 */
export function WidgetConfig() { }
if (false) {
    /** @type {?|undefined} */
    WidgetConfig.prototype.name;
    /**
     * 是否在部件配置中开启数据配置、筛选配置，默认 false
     * @type {?|undefined}
     */
    WidgetConfig.prototype.enableDataConfig;
    /**
     * 是否使用公共的标题头，默认 false
     * @type {?|undefined}
     */
    WidgetConfig.prototype.enableCommonTitleBar;
    /**
     * 是否启用加载 assets/i18n/ 下的国际化资源文件，默认 false
     * @type {?|undefined}
     */
    WidgetConfig.prototype.enableLoadingI18n;
    /**
     * 自定义国际化资源文件目录位置，如 '/apps/test/i18n/'，默认位置为元部件自身目录下的 assets/i18n/ 位置。
     * @type {?|undefined}
     */
    WidgetConfig.prototype.customI18nResourceUrl;
}
/**
 * @record
 */
export function RegisteredWidgetConfig() { }
if (false) {
    /** @type {?|undefined} */
    RegisteredWidgetConfig.prototype.ctor;
}
