// @gsp-svc/expression/esm2015/lib/expression-schema-grid/expression-schema-grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { CurrentNode } from '../currentNode';
export class ExpressionSchemaGridComponent {
    constructor() {
        this.schema = new Map();
        this.selectschemachange = new EventEmitter();
        this.currentNode = new CurrentNode;
        this.cols = [
            { field: 'name', title: '纲要名称' }
        ];
        this.treedata = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.convertSchema();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    selectRowChange($event) {
        this.currentNode = new CurrentNode;
        this.currentNode.id = $event.node.data.id;
        this.currentNode.name = $event.node.data.name;
        this.currentNode.parentName = $event.node.parent.data.name;
        this.selectschemachange.emit(this.currentNode);
    }
    /**
     * @return {?}
     */
    convertSchema() {
        this.schema.forEach((/**
         * @param {?} value
         * @param {?} key
         * @return {?}
         */
        (value, key) => {
            /** @type {?} */
            let treeitem = {
                data: {
                    "id": Math.random(),
                    "name": "",
                    "type": "",
                    "description": ""
                },
                children: []
            };
            treeitem.data.name = key;
            treeitem.data.type = "Schema";
            treeitem.data.description = value.description;
            value.entityTypes.forEach((/**
             * @param {?} entitytype
             * @return {?}
             */
            entitytype => {
                /** @type {?} */
                let typetree = {
                    data: {
                        "id": "",
                        "name": "",
                        "type": "",
                        "description": ""
                    },
                    children: [],
                    parent: treeitem
                };
                typetree.data.id = entitytype.id;
                typetree.data.name = entitytype.name;
                typetree.data.type = "EntityType";
                typetree.data.description = entitytype.description;
                treeitem.children.push(typetree);
            }));
            this.treedata.push(treeitem);
        }));
    }
}
ExpressionSchemaGridComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-expression-schema-grid',
                template: "<farris-treetable #tt \r\n[data]=\"treedata\"\r\n[fit]=\"true\" \r\n[columns]=\"cols\" \r\n[idField]=\"'id'\"\r\n[height]=\"268\"\r\n[showFilterBar]=\"false\"\r\n[showHeader]=\"true\"\r\n[enableFindText]=\"true\"\r\n(cellClick)=\"selectRowChange($event)\"\r\n>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
ExpressionSchemaGridComponent.ctorParameters = () => [];
ExpressionSchemaGridComponent.propDecorators = {
    schema: [{ type: Input }],
    selectschemachange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    ExpressionSchemaGridComponent.prototype.schema;
    /** @type {?} */
    ExpressionSchemaGridComponent.prototype.selectschemachange;
    /** @type {?} */
    ExpressionSchemaGridComponent.prototype.currentNode;
    /** @type {?} */
    ExpressionSchemaGridComponent.prototype.cols;
    /** @type {?} */
    ExpressionSchemaGridComponent.prototype.treedata;
}
