// @gsp-wf/wf-process-editor/esm5/lib/app/auto-activity.adapter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ProcessDesignerUIState } from './process-designer.uistate';
import { ActivityComponentComponent, ActivityComponentConverter } from '../views/activity-component/activity-component.component';
import { VariableSettingComponent, VariableSettingConverter } from '../views/variable-setting/variable-setting.component';
import { AutoActivityPropertyKeys } from '../designer/wf-cell/property-keys/AutoActivityPropertyKeys';
import { PerformerOptionsComponent, PerformerOptionsConverter } from '../views/performer/views/performer-options/performer-options.component';
import { WfLocalePipe } from '../i18n/pipe/wf-process-editor-locale.pipe';
var AutoActivityAdapter = /** @class */ (function () {
    function AutoActivityAdapter(uiState, localPipe) {
        this.uiState = uiState;
        this.localPipe = localPipe;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    AutoActivityAdapter.prototype.setProperty = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        this.uiState.propertyData = obj.propertyData;
        this.uiState.propertyConfig = this.setConfig(obj.propertyConfig);
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    AutoActivityAdapter.prototype.setConfig = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        /** @type {?} */
        var propertyConfig = [];
        if (obj) {
            obj.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                _this.setEntity(element.properties);
                propertyConfig.push(element);
            }));
        }
        return propertyConfig;
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    AutoActivityAdapter.prototype.setEntity = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        if (obj && obj.length > 0) {
            obj.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                switch (item.propertyID) {
                    case AutoActivityPropertyKeys.Tools: // 参与者
                        item.editor = ActivityComponentComponent;
                        item.converter = new ActivityComponentConverter(_this.localPipe);
                        item.editorParams = {
                            location: 'autoActivity'
                        };
                        break;
                    case AutoActivityPropertyKeys.variableSettings: // 变量设置
                        item.editor = VariableSettingComponent;
                        item.converter = new VariableSettingConverter(_this.localPipe);
                        break;
                    case AutoActivityPropertyKeys.ParticipantOptions: // 参与者选项
                        item.editor = PerformerOptionsComponent;
                        item.converter = new PerformerOptionsConverter(_this.localPipe);
                        item.editorParams = {
                            location: 'autoActivity'
                        };
                        break;
                    default:
                        break;
                }
            }));
        }
    };
    AutoActivityAdapter.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AutoActivityAdapter.ctorParameters = function () { return [
        { type: ProcessDesignerUIState },
        { type: WfLocalePipe }
    ]; };
    return AutoActivityAdapter;
}());
export { AutoActivityAdapter };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AutoActivityAdapter.prototype.uiState;
    /**
     * @type {?}
     * @private
     */
    AutoActivityAdapter.prototype.localPipe;
}
