// @gsp-wf/wf-process-editor/esm5/lib/app/bill-node-activity.adapter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ProcessDesignerUIState } from './process-designer.uistate';
import { VoConfigComponent, VoConfigConverter } from '../views/vo-config/vo-config.component';
import { WfOptionsComponent, WfOptionsConverter } from '../views/wf-options/wf-options.component';
import { WfLocalePipe } from '../i18n/pipe/wf-process-editor-locale.pipe';
var BillNodeActivityAdapter = /** @class */ (function () {
    function BillNodeActivityAdapter(uiState, localPipe) {
        this.uiState = uiState;
        this.localPipe = localPipe;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    BillNodeActivityAdapter.prototype.setProperty = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        this.uiState.propertyData = obj.propertyData;
        this.uiState.propertyConfig = this.setConfig(obj.propertyConfig);
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    BillNodeActivityAdapter.prototype.setConfig = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        /** @type {?} */
        var propertyConfig = [];
        if (obj) {
            obj.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                _this.setEntity(element.properties);
                propertyConfig.push(element);
            }));
        }
        return propertyConfig;
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    BillNodeActivityAdapter.prototype.setEntity = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        if (obj && obj.length > 0) {
            obj.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                switch (item.propertyID) {
                    case "VoConfig":
                        item.converter = new VoConfigConverter(_this.localPipe);
                        item.editor = VoConfigComponent;
                        item.editorParams = {
                            bizTypeId: _this.uiState.flowFormInfo.flowForm.businessObjectId
                        };
                        break;
                    case "WfOpts":
                        item.converter = new WfOptionsConverter(_this.localPipe);
                        item.editor = WfOptionsComponent;
                        item.editorParams = {
                            bizTypeId: _this.uiState.flowFormInfo.flowForm.businessObjectId
                        };
                        break;
                    case "wfConfig": //启用工作流
                        item.iterator = _this.uiState.boolIterator;
                        break;
                    default:
                        break;
                }
            }));
        }
    };
    BillNodeActivityAdapter.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    BillNodeActivityAdapter.ctorParameters = function () { return [
        { type: ProcessDesignerUIState },
        { type: WfLocalePipe }
    ]; };
    return BillNodeActivityAdapter;
}());
export { BillNodeActivityAdapter };
if (false) {
    /**
     * @type {?}
     * @private
     */
    BillNodeActivityAdapter.prototype.uiState;
    /**
     * @type {?}
     * @private
     */
    BillNodeActivityAdapter.prototype.localPipe;
}
