// @gsp-wf/wf-process-editor/esm5/lib/app/catch-event.adapter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ProcessDesignerUIState } from './process-designer.uistate';
import { Injectable, Injector } from '@angular/core';
import { BackwardRuleComponent, BackwardRuleConverter } from '../views/backward-rule/backward-rule.component';
import { EventDefinitionComponent, EventDefinitionConverter } from '../views/event-definition/event-definition.component';
import { VariableSettingComponent, VariableSettingConverter } from '../views/variable-setting/variable-setting.component';
import { IntermediateCatchEventImplPropertyKeys } from '../designer/wf-cell/property-keys/IntermediateCatchEventImplPropertyKeys';
import { WfLocalePipe } from '../i18n/pipe/wf-process-editor-locale.pipe';
var CatchEventAdapter = /** @class */ (function () {
    function CatchEventAdapter(uiState, injector, localPipe) {
        this.uiState = uiState;
        this.injector = injector;
        this.localPipe = localPipe;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    CatchEventAdapter.prototype.setProperty = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        this.uiState.propertyData = obj.propertyData;
        this.uiState.propertyConfig = this.setConfig(obj.propertyConfig);
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    CatchEventAdapter.prototype.setConfig = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        /** @type {?} */
        var propertyConfig = [];
        if (obj) {
            obj.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                _this.setEntity(element.properties);
                propertyConfig.push(element);
            }));
        }
        return propertyConfig;
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    CatchEventAdapter.prototype.setEntity = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        if (obj && obj.length > 0) {
            obj.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                switch (item.propertyID) {
                    case IntermediateCatchEventImplPropertyKeys.BackWardRule:
                        item.editor = BackwardRuleComponent;
                        item.converter = new BackwardRuleConverter(_this.localPipe);
                        item.editorParams = {
                            nodeType: 'CallActivitySubProcess'
                        };
                        break;
                    case IntermediateCatchEventImplPropertyKeys.EventDefinitions:
                        item.editor = EventDefinitionComponent;
                        item.converter = new EventDefinitionConverter(_this.localPipe);
                        break;
                    case IntermediateCatchEventImplPropertyKeys.variableSettings: // 变量设置
                        item.editor = VariableSettingComponent;
                        item.converter = new VariableSettingConverter(_this.localPipe);
                        break;
                    default:
                        break;
                }
            }));
        }
    };
    CatchEventAdapter.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CatchEventAdapter.ctorParameters = function () { return [
        { type: ProcessDesignerUIState },
        { type: Injector },
        { type: WfLocalePipe }
    ]; };
    return CatchEventAdapter;
}());
export { CatchEventAdapter };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CatchEventAdapter.prototype.uiState;
    /**
     * @type {?}
     * @private
     */
    CatchEventAdapter.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    CatchEventAdapter.prototype.localPipe;
}
