// @farris/ui-feature-editor/esm2015/lib/feature-editor.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { FarrisFormsModule } from '@farris/ui-forms';
import { LookupModule } from '@farris/ui-lookup';
import { ComboListModule } from '@farris/ui-combo-list';
import { InputGroupModule } from '@farris/ui-input-group';
import { ModalModule } from '@farris/ui-modal';
import { FarrisDatePickerModule } from '@farris/ui-datepicker';
import { NumberSpinnerModule } from '@farris/ui-number-spinner';
import { TimePickerModule } from '@farris/ui-time-picker';
import { LocaleModule } from '@farris/ui-locale';
import { SwitchModule } from '@farris/ui-switch';
import { FarrisSectionModule } from '@farris/ui-section';
import { FeatureEditorComponent } from './feature-editor.component';
import { LookupDirective } from './lookup.directive';
import { I18nService } from './i18n-service';
export class FeatureEditorModule {
}
FeatureEditorModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    FeatureEditorComponent,
                    LookupDirective
                ],
                imports: [
                    CommonModule,
                    FormsModule,
                    FarrisFormsModule,
                    LookupModule,
                    ComboListModule,
                    InputGroupModule,
                    ModalModule,
                    FarrisDatePickerModule,
                    NumberSpinnerModule,
                    TimePickerModule,
                    SwitchModule,
                    FarrisSectionModule,
                    LocaleModule.forRoot(),
                ],
                exports: [
                    FeatureEditorComponent,
                    LookupDirective
                ],
                entryComponents: [
                    FeatureEditorComponent
                ],
                providers: [
                    I18nService
                ]
            },] }
];
