// @farris/ui-feature-editor/esm2015/lib/type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 特性属性
 */
// export interface IFeatures {
//   id: string;
//   /**
//    * 编号
//    */
//   code: string;
//   /**
//    * 名称
//    */
//   __name__: { "zh-CHS": string, "zh-CHT": string, en: string };
//   name_chs: string;
//   name_cht: string;
//   name_en: string;
//   /**
//    * 是否必填
//    * @description 可选值 "1" | "0"
//    */
//   isrequired: string;
//   /**
//    * 输入控件类型
//    * @description 可选值 Help | Number | Enum etc
//    */
//   inputtype: string;
//   textfield: string;
//   codefield: string;
//   /**
//    * 过滤条件
//    */
//   filtercondition: string;
//   /**
//    * 枚举信息
//    */
//   enuminfo: string;
//   /**
//    * 枚举数据源
//    */
//   __data__: Array<{ id: string, text: string }>;
//   /**
//    * 精度
//    */
//   digitalaccuracy: number;
//   /**
//    * 特征集id
//    */
//   setid: string;
//   /**
//    * 是否是树帮助
//    * @description 可选值 "0" | "1"
//    */
//   setistree: string;
//   /**
//    * idp帮助id
//    */
//   codesetid: string;
//   /**
//    * 只读
//    * @description 可选值 "0" | "1"
//    */
//   isreadonly: string;
//   /**
//    * 顺序号
//    */
//   sortorder: number;
//   value: string;
//   helpid: string;
/**
 * @record
 */
export function IFeatureEditorOptions() { }
if (false) {
    /** @type {?|undefined} */
    IFeatureEditorOptions.prototype.width;
    /** @type {?|undefined} */
    IFeatureEditorOptions.prototype.height;
    /** @type {?|undefined} */
    IFeatureEditorOptions.prototype.title;
    /**
     * 确认事件
     * @type {?|undefined}
     */
    IFeatureEditorOptions.prototype.onConfirm;
    /**
     * 确认前
     * \@description 方法返回值可以为
     * @type {?|undefined}
     */
    IFeatureEditorOptions.prototype.onConfirming;
    /**
     * 特性编辑器打开前
     * @type {?|undefined}
     */
    IFeatureEditorOptions.prototype.onOpening;
}
/** @enum {string} */
const InputType = {
    Help: 'Help',
    String: 'String',
    Enum: 'Enum',
    Date: 'Date',
};
export { InputType };
/** @enum {string} */
const EditorTypes = {
    TEXTAREA: 'textarea',
    TEXTBOX: 'textbox',
    CHECKBOX: 'checkbox',
    DATEPICKER: 'datepicker',
    INPUTGROUP: 'input-group',
    SELECT: 'select',
    LOOKUP: 'lookup',
    NUMBERBOX: 'numberbox',
    COMBOLIST: 'combolist',
    SWITCH: 'switch',
    TIMEPICKER: 'timepicker',
    COMBOLOOKUP: 'combo-lookup',
    LANGUAGETEXTBOX: 'language-textbox',
    ENUMEDITOR: 'enum-editor',
    FILTEREDITOR: 'filter-editor',
    SORTEDITOR: 'sort-editor',
};
export { EditorTypes };
;
/**
 * @record
 */
export function IEditor() { }
if (false) {
    /** @type {?} */
    IEditor.prototype.controlType;
    /* Skipping unhandled member: [prop: string]: any;*/
}
/**
 * 枚举编辑器
 * @record
 */
export function IEnumEditor() { }
if (false) {
    /** @type {?} */
    IEnumEditor.prototype.controlType;
    /** @type {?} */
    IEnumEditor.prototype.idField;
    /** @type {?} */
    IEnumEditor.prototype.textField;
    /** @type {?|undefined} */
    IEnumEditor.prototype.multiSelect;
    /** @type {?|undefined} */
    IEnumEditor.prototype.items;
}
;
/**
 * 数字编辑器
 * @record
 */
export function INumberEditor() { }
if (false) {
    /** @type {?} */
    INumberEditor.prototype.controlType;
    /** @type {?} */
    INumberEditor.prototype.precision;
    /** @type {?|undefined} */
    INumberEditor.prototype.step;
    /** @type {?|undefined} */
    INumberEditor.prototype.min;
    /** @type {?|undefined} */
    INumberEditor.prototype.max;
}
;
/**
 * 日期编辑器
 * @record
 */
export function IDateEditor() { }
if (false) {
    /** @type {?} */
    IDateEditor.prototype.controlType;
    /** @type {?|undefined} */
    IDateEditor.prototype.dateRange;
}
/**
 * 时间编辑器
 * @record
 */
export function ITimeEditor() { }
if (false) {
    /** @type {?} */
    ITimeEditor.prototype.controlType;
    /** @type {?|undefined} */
    ITimeEditor.prototype.use12Hours;
}
/**
 * 帮助编辑器
 * @record
 */
export function ILookupEditor() { }
if (false) {
    /** @type {?} */
    ILookupEditor.prototype.idField;
    /** @type {?} */
    ILookupEditor.prototype.textField;
    /** @type {?} */
    ILookupEditor.prototype.valueField;
    /** @type {?|undefined} */
    ILookupEditor.prototype.displayType;
    /** @type {?|undefined} */
    ILookupEditor.prototype.singleSelect;
    /** @type {?|undefined} */
    ILookupEditor.prototype.multiSelect;
    /** @type {?|undefined} */
    ILookupEditor.prototype.mapFields;
    /** @type {?|undefined} */
    ILookupEditor.prototype.dialogTitle;
    /** @type {?|undefined} */
    ILookupEditor.prototype.hiddenInputName;
}
/**
 * 特性描述
 * @record
 */
export function IFeature() { }
if (false) {
    /** @type {?} */
    IFeature.prototype.id;
    /** @type {?} */
    IFeature.prototype.title;
    /** @type {?} */
    IFeature.prototype.field;
    /** @type {?} */
    IFeature.prototype.editor;
    /** @type {?|undefined} */
    IFeature.prototype.value;
    /** @type {?|undefined} */
    IFeature.prototype.results;
    /** @type {?|undefined} */
    IFeature.prototype.options;
}
