// @gsp-wf/rtdevkit/esm5/lib/runtime/components/task-assign/task-assign.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { TaskService } from '../../services/task.service';
import { UserService } from '../../services/user.service';
import { UserHelperComponent } from '../user-helper/user-helper.component';
import { NotifyService } from '@farris/ui-notify';
import { I18nService } from '../../services/i18n.service';
var TaskAssignComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaskAssignComponent, _super);
    function TaskAssignComponent(taskSvc, userSvc, notify, i18nService) {
        var _this = _super.call(this, i18nService, userSvc) || this;
        _this.taskSvc = taskSvc;
        _this.userSvc = userSvc;
        _this.notify = notify;
        _this.i18nService = i18nService;
        _this.completed = new EventEmitter();
        _this.paginationOptions.id = 'WF-TaskAss-DataTable-Pagination';
        return _this;
        // this.title = this.i18nService.getValue('static.assignee.title');
    }
    Object.defineProperty(TaskAssignComponent.prototype, "assigneeInfo", {
        // title = '请选择指派用户';
        set: 
        // title = '请选择指派用户';
        /**
         * @param {?} assigneeInfos
         * @return {?}
         */
        function (assigneeInfos) {
            if (assigneeInfos && assigneeInfos.length > 0) {
                this.assignees = assigneeInfos;
                this.showModal(true);
                this.execute();
            }
            else {
                this.showModal(false);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TaskAssignComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    TaskAssignComponent.prototype.execute = /**
     * @return {?}
     */
    function () {
        this.assigneesPa = {
            workItemId: this.assignees[0].workItemId,
            assigneeIds: [],
            comment: ''
        };
        this.singleSelect = !this.assignees[0].multiSelect;
        this.allowGlobalAssign = this.assignees[0].allowGlobalAssign;
        if (this.assignees[0].assigneeUsers && this.assignees[0].assigneeUsers.length) {
            this.getAssignedUsers();
        }
        else {
            if (this.allowGlobalAssign) {
                this.getUserList();
            }
        }
    };
    /**
     * @return {?}
     */
    TaskAssignComponent.prototype.getAssignedUsers = /**
     * @return {?}
     */
    function () {
        this.allUserData = this.assignees[0].assigneeUsers.map((/**
         * @param {?} user
         * @return {?}
         */
        function (user) {
            return {
                ID: user.id,
                Name: user.name,
                Code: user.code,
                OrgId: user.orgId,
                OrgName: user.orgNamePath
            };
        }));
        this.userListData = this.allUserData;
    };
    /**
     * @return {?}
     */
    TaskAssignComponent.prototype.handleOk = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.selectedUsers = this.singleSelect ? [this.userTable.selections] : this.userTable.selections;
        if (!this.selectedUsers || this.selectedUsers.length < 1) {
            this.notify.warning(this.i18nService.getValue('static.assignee.noAssignee'));
            return;
        }
        this.assignees.shift();
        this.assigneesPa.assigneeIds = this.selectedUsers.map((/**
         * @param {?} user
         * @return {?}
         */
        function (user) { return user.ID; }));
        // 检查指定是否完成，如果完成提交，否则继续指派任务
        if (this.assignees.length > 0) {
            this.userListData = []; // clear source
            this.execute();
        }
        else {
            this.taskSvc.assignNextUsersByWorkItemId(this.assigneesPa).subscribe((/**
             * @return {?}
             */
            function () {
                _this.showModal(false);
                _this.notify.success(_this.i18nService.getValue('static.assignee.successAssign'));
                _this.completed.emit(true);
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) {
                if (error && error.error && error.error.Message) {
                    _this.notify.error(error.error.Message);
                }
            }));
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TaskAssignComponent.prototype.search = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        if (!this.assignees[0].assigneeUsers || !this.assignees[0].assigneeUsers.length) {
            _super.prototype.search.call(this, event);
        }
        else {
            this.searchValue = event.value;
            if (this.searchValue) {
                this.userListData = this.allUserData.filter((/**
                 * @param {?} user
                 * @return {?}
                 */
                function (user) {
                    return (user.Name) && user.Name.includes(_this.searchValue) ||
                        (user.Code) && user.Code.includes(_this.searchValue);
                }));
            }
            else {
                this.getAssignedUsers();
            }
        }
    };
    /**
     * @protected
     * @param {?} show
     * @return {?}
     */
    TaskAssignComponent.prototype.showModal = /**
     * @protected
     * @param {?} show
     * @return {?}
     */
    function (show) {
        if (show) {
            this.userDialog.show();
        }
        else {
            // 临时处理方案
            if (this.userDialog.beforeClose) {
                this.userDialog.close();
            }
        }
    };
    /**
     * @param {?} pageSize
     * @return {?}
     */
    TaskAssignComponent.prototype.onPageSizeChange = /**
     * @param {?} pageSize
     * @return {?}
     */
    function (pageSize) {
        this.paginationOptions.itemsPerPage = pageSize;
        if (this.allowGlobalAssign) {
            this.getUserList();
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TaskAssignComponent.prototype.onPageChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.paginationOptions.currentPage = event.pageIndex;
        if (this.allowGlobalAssign) {
            this.getUserList();
        }
    };
    TaskAssignComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-task-assign',
                    template: "<farris-dialog #allUserDialog [buttons]=\"buttonRef\" [width]=\"850\" [height]=\"520\" [title]=\"'\u7CFB\u7EDF\u7528\u6237'\">\r\n  <div class=\"d-flex flex-column my-data-grid h-100\" style=\"padding: 0 14px;\">\r\n    <div class=\"flex-auto\" style=\"position:relative;\">\r\n      <farris-datatable #allUserTable [columns]=\"columns\" [data]=\"userListData | paginate: paginationOptions\"\r\n        [showFilterBar]='true' [singleSelect]=\"singleSelect\" [idField]=\"'ID'\" [remote]=\"'server'\" (search)=\"search($event)\"\r\n        [pagination]='false'></farris-datatable>\r\n    </div>\r\n    <div style=\"position:relative;\">\r\n      <pagination-controls [id]=\"paginationOptions.id\" [previousLabel]=\"' '\" [nextLabel]=\"' '\"\r\n        (pageChange)=\"onPageChange($event)\" (pageSizeChange)=\"onPageSizeChange($event)\">\r\n      </pagination-controls>\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n<ng-template #buttonRef>\r\n  <button class=\"btn btn-sm btn-primary px-3\" (click)=\"handleOk()\">\u786E\u5B9A</button>\r\n  <button class=\"btn btn-sm btn-default px-3\" (click)=\"handleCancel()\">\u53D6\u6D88</button>\r\n</ng-template>",
                    styles: [".my-data-grid .filter-row .search-fields{padding-left:0!important}.my-data-grid .filter-row .search-text{padding-right:0!important}"]
                }] }
    ];
    /** @nocollapse */
    TaskAssignComponent.ctorParameters = function () { return [
        { type: TaskService },
        { type: UserService },
        { type: NotifyService },
        { type: I18nService }
    ]; };
    TaskAssignComponent.propDecorators = {
        assigneeInfo: [{ type: Input }],
        completed: [{ type: Output }]
    };
    return TaskAssignComponent;
}(UserHelperComponent));
export { TaskAssignComponent };
if (false) {
    /** @type {?} */
    TaskAssignComponent.prototype.completed;
    /**
     * @type {?}
     * @private
     */
    TaskAssignComponent.prototype.allUserData;
    /**
     * @type {?}
     * @private
     */
    TaskAssignComponent.prototype.assignees;
    /**
     * @type {?}
     * @private
     */
    TaskAssignComponent.prototype.assigneesPa;
    /**
     * @type {?}
     * @private
     */
    TaskAssignComponent.prototype.allowGlobalAssign;
    /** @type {?} */
    TaskAssignComponent.prototype.taskSvc;
    /**
     * @type {?}
     * @protected
     */
    TaskAssignComponent.prototype.userSvc;
    /**
     * @type {?}
     * @protected
     */
    TaskAssignComponent.prototype.notify;
    /**
     * @type {?}
     * @protected
     */
    TaskAssignComponent.prototype.i18nService;
}
