// @gspwidget/util/esm5/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of util
 */
export { UtilService } from './lib/util.service';
export { UtilComponent } from './lib/util.component';
export { UtilModule } from './lib/util.module';
export { backendString2Json, list2Tree, uuidv4, appendScript, loadCss, deepClone, setArrayLengthTo, getCurTheme, loadCommonStyleNTheme, objectLiteralEqual, curLang } from './lib/misc';
export { addBaseUrl, LS_LOAD_LOCAL_ASSET, IGIX_HOME_PATH, BASE_URL, DATA_PREFIX, FILTERS_WIDGET_NAME, FILTERS_WIDGET_FILTERS_PROP_NAME, FILTERS_WIDGET_PROP_NAME, DESIGNER_CONFIG, dimensionTypeConfig } from './lib/const';
export { IgixUtilService } from './lib/igix-util.service';
export { onRtfTabChange, RtfUtilService } from './lib/rtf-util.service';
export { portletUrls, portletOps, PortletRepo } from './lib/repositories/portlet.repo';
export { convertWidget, widgetUrls, WidgetRepo } from './lib/repositories/widget.repo';
export { datasetUrls, aggUrls, dataResultSetUrls, DatasetRepo } from './lib/repositories/dataset.repo';
export { datasrcUrls, DatasrcRepo } from './lib/repositories/datasrc.repo';
export { igixUrls, IgixRepo } from './lib/repositories/igix.repo';
export { convertPageInfo, convertPageInfoInverse, pageInfoUrls, pageRoles, PageRepo } from './lib/repositories/page.repo';
export { Lang, I18nValue } from './lib/types/common.type';
export { Datasrc, DataParam, DatasetParameter, Dataset, Dimension, ComputedType, AggType, ComputedCol, Agg, AggregationResultSetType, DpValue, Indicator, RuntimeAggOption } from './lib/types/data.type';
export { FuncType } from './lib/types/igix.type';
export { PageInfo, PageConfig } from './lib/types/page.type';
export { DEFAULT_PORTLET_COLS, DEFAULT_PORTLET_ROWS, Portlet, EppPortlet, BAPortlet } from './lib/types/portlet.type';
export { WidgetContainerMode } from './lib/types/widget.type';
export { encrypt, decrypt } from './lib/utils/encryption';
export { mapEntityFeilds } from './lib/utils/field-mapping';
export { TransPipe } from './lib/trans.pipe';
export { ArrUtil } from './lib/arr';
export { PlThemeColors } from './lib/colors';
