// @gsp-wf/wf-process-editor/esm2015/lib/views/performer/views/reporting-line/service/wf-report-def-help.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { map } from 'rxjs/operators';
export class WfReportDefHelpService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @param {?} url
     * @param {?=} params
     * @return {?}
     */
    getData(url, params) {
        /** @type {?} */
        const customData = params.customData;
        /** @type {?} */
        const index = params.pageIndex || 1;
        /** @type {?} */
        const size = params.pageSize || 20;
        /** @type {?} */
        const searchValue = JSON.parse(params.searchValue);
        if (searchValue.searchValue || customData.filter) {
            /** @type {?} */
            let query = '{';
            if (searchValue.searchValue) {
                if (searchValue.searchField === '*') {
                    query += `"codeOrNameLike":"${searchValue.searchValue}",`;
                }
                else {
                    query += `"${searchValue.searchField}":"${searchValue.searchValue}",`;
                }
            }
            if (customData.filter) {
                /** @type {?} */
                const keys = Object.keys(customData.filter);
                if (keys && keys.length > 0) {
                    for (const key of keys) {
                        if (customData.filter[key]) {
                            query += `"${key}":"${customData.filter[key]}",`;
                        }
                    }
                }
            }
            if (query.substr(query.length - 1, 1) === ',') {
                query = query.substring(0, query.length - 1);
            }
            query += '}';
            query = encodeURIComponent(query);
            query += `&pageIndex=${index}&pageSize=${size}`;
            url += '?pageFilter=' + query;
        }
        else {
            url += `?pageIndex=${index}&pageSize=${size}`;
        }
        return this.http.get(url).pipe(map((/**
         * @param {?} re
         * @return {?}
         */
        (re) => ({
            items: re.content,
            total: re.totalElements,
            pageInfo: {
                pageIndex: re.pageable.pageNumber + 1,
                pageSize: re.pageable.pageSize,
                pageList: [20, 40, 60, 80, 100],
                enablePager: true
            }
        }))));
    }
}
WfReportDefHelpService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
WfReportDefHelpService.ctorParameters = () => [
    { type: HttpService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    WfReportDefHelpService.prototype.http;
}
