// @edp-bif/runtime-api/esm5/lib/entity/form-rest/FormGetBillCreationProcessEntityExArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/form-rest/FormGetBillCreationProcessEntityExArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from "@edp-aif/common-api";
import { ManualCreationMode } from './ManualCreationMode';
var FormGetBillCreationProcessEntityExArgs = /** @class */ (function (_super) {
    tslib_1.__extends(FormGetBillCreationProcessEntityExArgs, _super);
    function FormGetBillCreationProcessEntityExArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormGetBillCreationProcessEntityExArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.upBillVoId = null;
        this.downBillVoId = null;
        this.manualCreationMode = null;
    };
    /**
     * @return {?}
     */
    FormGetBillCreationProcessEntityExArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoId", this.upBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoId", this.downBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "manualCreationMode", this.manualCreationMode);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormGetBillCreationProcessEntityExArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.upBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoId");
        this.downBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoId");
        this.manualCreationMode = ManualCreationMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "manualCreationMode")];
    };
    return FormGetBillCreationProcessEntityExArgs;
}(JsonConvertableImpl));
export { FormGetBillCreationProcessEntityExArgs };
if (false) {
    /** @type {?} */
    FormGetBillCreationProcessEntityExArgs.prototype.upBillVoId;
    /** @type {?} */
    FormGetBillCreationProcessEntityExArgs.prototype.downBillVoId;
    /** @type {?} */
    FormGetBillCreationProcessEntityExArgs.prototype.manualCreationMode;
}
