// @edp-bif/runtime-api/esm5/lib/entity/form-rest/GeneralTextIdWithDataId.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/form-rest/GeneralTextIdWithDataId.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
var GeneralTextIdWithDataId = /** @class */ (function (_super) {
    tslib_1.__extends(GeneralTextIdWithDataId, _super);
    function GeneralTextIdWithDataId() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    GeneralTextIdWithDataId.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.generalTextId = null;
        this.dataId = null;
    };
    /**
     * @return {?}
     */
    GeneralTextIdWithDataId.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "generalTextId", this.generalTextId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dataId", this.dataId);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    GeneralTextIdWithDataId.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.generalTextId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "generalTextId");
        this.dataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dataId");
    };
    return GeneralTextIdWithDataId;
}(JsonConvertableImpl));
export { GeneralTextIdWithDataId };
if (false) {
    /** @type {?} */
    GeneralTextIdWithDataId.prototype.generalTextId;
    /** @type {?} */
    GeneralTextIdWithDataId.prototype.dataId;
}
