// @gsp-sys/sysmgr-ui/esm5/lib/func-op-help/biz-object/biz-object.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { BizObjectService } from '../service/biz-object.service';
var BizObjectComponent = /** @class */ (function () {
    function BizObjectComponent(bizObjectService) {
        this.bizObjectService = bizObjectService;
        this.selectKeys = [];
        /**
         * 业务对象变化后
         */
        this.bizObjectChanged = new EventEmitter();
        /**
         * 树控件展开的节点值
         */
        this.expandedKeys = [];
    }
    /**
     * @return {?}
     */
    BizObjectComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.bizObjectService.getBizObjects().subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            result.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                if (!element.parentID) {
                    element.parentID = undefined;
                    // 展开第一级
                    if (element.layer === 1) {
                        _this.expandedKeys.push(element.id);
                    }
                }
            }));
            _this.data = result;
            // 默认业务对象的处理
            if (_this.bizObjectId && _this.bizObjectId !== '') {
                // 树焦点行
                /** @type {?} */
                var curBizObjectId_1 = _this.bizObjectId;
                while (curBizObjectId_1) {
                    _this.expandedKeys.push(curBizObjectId_1);
                    // 找到当前节点的父节点
                    curBizObjectId_1 = _this.data.find((/**
                     * @param {?} item
                     * @return {?}
                     */
                    function (item) { return item.id === curBizObjectId_1; })).parentID;
                }
                // 右侧操作变化
                _this.selectKeys = [_this.bizObjectId];
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    _this.bizObjectChanged.emit(_this.bizObjectId);
                }), 80);
            }
        }));
    };
    /** 选中行改变后触发事件 */
    /**
     * 选中行改变后触发事件
     * @param {?} args
     * @return {?}
     */
    BizObjectComponent.prototype.selectChange = /**
     * 选中行改变后触发事件
     * @param {?} args
     * @return {?}
     */
    function (args) {
        /** @type {?} */
        var curBusId = args.dataItem.id;
        // 首次触发
        if (this.selectKeys == null || this.selectKeys.length === 0) {
            this.bizObjectChanged.emit(curBusId);
        }
        else {
            // 这个事件里还不会对selectedkeys数组赋值，是上一个状态的
            /** @type {?} */
            var lastBusId = this.selectKeys[0];
            // 变化的情况下触发
            if (lastBusId !== curBusId) {
                this.bizObjectChanged.emit(curBusId);
            }
        }
    };
    BizObjectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-biz-object',
                    template: "<div class=\"d-flex flex-column flex-fill pt-2 pb-2\" style=\"height: 100%\">\r\n    <kendo-treeview [nodes]=\"data\" textField=\"name\" kendoTreeViewExpandable kendoTreeViewFlatDataBinding idField=\"id\"\r\n      parentIdField=\"parentID\" kendoTreeViewSelectable [selectBy]=\"'id'\" [(selectedKeys)]=\"this.selectKeys\"\r\n      [(expandedKeys)]=\"expandedKeys\"\r\n      [expandBy]=\"'id'\"\r\n      (selectionChange)=\"selectChange($event)\">\r\n    </kendo-treeview>\r\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    BizObjectComponent.ctorParameters = function () { return [
        { type: BizObjectService }
    ]; };
    BizObjectComponent.propDecorators = {
        bizObjectId: [{ type: Input }],
        bizObjectChanged: [{ type: Output }]
    };
    return BizObjectComponent;
}());
export { BizObjectComponent };
if (false) {
    /** @type {?} */
    BizObjectComponent.prototype.bizObjectId;
    /** @type {?} */
    BizObjectComponent.prototype.data;
    /** @type {?} */
    BizObjectComponent.prototype.selectKeys;
    /** @type {?} */
    BizObjectComponent.prototype.currentid;
    /**
     * 业务对象变化后
     * @type {?}
     */
    BizObjectComponent.prototype.bizObjectChanged;
    /**
     * 树控件展开的节点值
     * @type {?}
     */
    BizObjectComponent.prototype.expandedKeys;
    /**
     * @type {?}
     * @private
     */
    BizObjectComponent.prototype.bizObjectService;
}
