// @progress/kendo-angular-layout/dist/es2015/common/draggable.directive.js
import { EventEmitter, ElementRef, Directive, NgZone } from '@angular/core';
import Draggable from '@telerik/kendo-draggable';
/**
 * @hidden
 */
export class DraggableDirective {
    constructor(zone, element) {
        this.zone = zone;
        this.kendo = {
            drag: new EventEmitter(),
            press: new EventEmitter(),
            release: new EventEmitter()
        };
        if (typeof document !== 'undefined') {
            this.draggable = new Draggable({
                drag: (e) => this.kendo.drag.next(e),
                press: (e) => this.kendo.press.next(e),
                release: (e) => this.kendo.release.next(e)
            });
            this.zone.runOutsideAngular(() => this.draggable.bindTo(element.nativeElement));
        }
    }
    ngOnDestroy() {
        if (typeof document !== 'undefined') {
            this.draggable.destroy();
        }
    }
}
DraggableDirective.decorators = [
    { type: Directive, args: [{
                outputs: ['kendo.press', 'kendo.drag', 'kendo.release'],
                selector: '[kendoLayoutDraggable]'
            },] },
];
/** @nocollapse */
DraggableDirective.ctorParameters = () => [
    { type: NgZone, },
    { type: ElementRef, },
];
