// @gsp-sys/rtf-common/esm5/lib/eventbus/frm-evenbus.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { EventBusPipe } from './frm-eventpipe';
var FrmEventBus = /** @class */ (function () {
    function FrmEventBus() {
        this.eventMap = new Map();
    }
    // 发送事件，通知订阅者接收消息。
    // 发送事件，通知订阅者接收消息。
    /**
     * @param {?} tokenValue
     * @param {?} eventArgs
     * @return {?}
     */
    FrmEventBus.prototype.post = 
    // 发送事件，通知订阅者接收消息。
    /**
     * @param {?} tokenValue
     * @param {?} eventArgs
     * @return {?}
     */
    function (tokenValue, eventArgs) {
        var e_1, _a;
        /** @type {?} */
        var eventPipeList = this.eventMap.get(tokenValue);
        if (!eventPipeList) {
            eventPipeList = new Array();
            this.eventMap.set(tokenValue, eventPipeList);
        }
        /** @type {?} */
        var eventPipe = eventPipeList.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.examByTargetToken(tokenValue); }));
        if (!eventPipe) {
            eventPipe = new EventBusPipe(tokenValue, eventPipeList, null);
        }
        try {
            for (var eventPipeList_1 = tslib_1.__values(eventPipeList), eventPipeList_1_1 = eventPipeList_1.next(); !eventPipeList_1_1.done; eventPipeList_1_1 = eventPipeList_1.next()) {
                var eventPipeDetail = eventPipeList_1_1.value;
                if (eventPipeDetail.matchEmitterToken(tokenValue)) {
                    eventPipeDetail.post(eventArgs);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (eventPipeList_1_1 && !eventPipeList_1_1.done && (_a = eventPipeList_1.return)) _a.call(eventPipeList_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    /**
     * @param {?} token
     * @param {?} sub
     * @return {?}
     */
    FrmEventBus.prototype.initialPipe = /**
     * @param {?} token
     * @param {?} sub
     * @return {?}
     */
    function (token, sub) {
        /** @type {?} */
        var pipes = this.eventMap.get(token);
        if (!pipes) {
            pipes = new Array();
            this.eventMap.set(token, pipes);
        }
        /** @type {?} */
        var eventPipe = pipes.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.examByTargetToken(token); }));
        if (!eventPipe) {
            eventPipe = new EventBusPipe(token, pipes, sub);
        }
    };
    // 订阅事件
    // 订阅事件
    /**
     * @param {?} tokenValue
     * @param {?} handler
     * @return {?}
     */
    FrmEventBus.prototype.on = 
    // 订阅事件
    /**
     * @param {?} tokenValue
     * @param {?} handler
     * @return {?}
     */
    function (tokenValue, handler) {
        /** @type {?} */
        var pipe = this.getEventPipe(tokenValue);
        if (pipe) {
            return pipe.subscribe(handler);
        }
    };
    /** 订阅事件，并记录订阅者唯一标识 */
    /**
     * 订阅事件，并记录订阅者唯一标识
     * @param {?} tokenValue
     * @param {?} observerKey
     * @param {?} handler
     * @return {?}
     */
    FrmEventBus.prototype.onWithObkey = /**
     * 订阅事件，并记录订阅者唯一标识
     * @param {?} tokenValue
     * @param {?} observerKey
     * @param {?} handler
     * @return {?}
     */
    function (tokenValue, observerKey, handler) {
        /** @type {?} */
        var pipe = this.getEventPipe(tokenValue);
        if (pipe) {
            return pipe.subscribeWithObkey(observerKey, handler);
        }
    };
    // 订阅一次。接收到一次消息之后自动取消订阅
    // 订阅一次。接收到一次消息之后自动取消订阅
    /**
     * @param {?} tokenValue
     * @param {?} handler
     * @return {?}
     */
    FrmEventBus.prototype.once = 
    // 订阅一次。接收到一次消息之后自动取消订阅
    /**
     * @param {?} tokenValue
     * @param {?} handler
     * @return {?}
     */
    function (tokenValue, handler) {
        /** @type {?} */
        var pipe = this.getEventPipe(tokenValue);
        if (pipe) {
            /** @type {?} */
            var res = pipe.subscribeOnce(handler);
            this.eventMap.delete(tokenValue);
            return res;
        }
    };
    /**
     * @param {?} tokenValue
     * @return {?}
     */
    FrmEventBus.prototype.unSubscribe = /**
     * @param {?} tokenValue
     * @return {?}
     */
    function (tokenValue) {
        /** @type {?} */
        var pipe = this.getEventPipe(tokenValue);
        if (pipe) {
            pipe.unSubscribe();
        }
    };
    /**
     * @private
     * @param {?} tokenValue
     * @return {?}
     */
    FrmEventBus.prototype.getEventPipe = /**
     * @private
     * @param {?} tokenValue
     * @return {?}
     */
    function (tokenValue) {
        /** @type {?} */
        var eventPipeList = this.eventMap.get(tokenValue);
        if (!eventPipeList) {
            return null;
        }
        /** @type {?} */
        var eventPipe = eventPipeList.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.examByTargetToken(tokenValue); }));
        return eventPipe;
    };
    FrmEventBus.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FrmEventBus.ctorParameters = function () { return []; };
    return FrmEventBus;
}());
export { FrmEventBus };
if (false) {
    /** @type {?} */
    FrmEventBus.prototype.eventMap;
}
