// @gsp-sys/rtf-common/esm5/lib/eventbus/frm-eventpipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// tslint:disable-next-line:import-blacklist
import { BehaviorSubject, Subscription, Subject } from 'rxjs';
import { Injectable } from '@angular/core';
var EventBusPipe = /** @class */ (function () {
    function EventBusPipe(tokenValue, parentEventPipeList, sub) {
        this.tokenValue = tokenValue;
        this.parentEventPipeList = parentEventPipeList;
        this.eventSubject = sub || new BehaviorSubject(null);
        this.subscription = new Subscription();
        this.onceSubscription = new Subscription();
        this.observers = new Map();
        if (this.parentEventPipeList) {
            this.parentEventPipeList.push(this);
        }
    }
    /**
     * @param {?} bizPara
     * @return {?}
     */
    EventBusPipe.prototype.post = /**
     * @param {?} bizPara
     * @return {?}
     */
    function (bizPara) {
        this.eventSubject.next(bizPara);
    };
    /**
     * @param {?} eventHandler
     * @return {?}
     */
    EventBusPipe.prototype.subscribe = /**
     * @param {?} eventHandler
     * @return {?}
     */
    function (eventHandler) {
        this.subscription = this.eventSubject.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            eventHandler(value);
        }));
        return this;
    };
    /**
     * @param {?} observerKey
     * @param {?} eventHandler
     * @return {?}
     */
    EventBusPipe.prototype.subscribeWithObkey = /**
     * @param {?} observerKey
     * @param {?} eventHandler
     * @return {?}
     */
    function (observerKey, eventHandler) {
        /** @type {?} */
        var observer = this.observers.get(observerKey);
        if (observer) {
            observer.unsubscribe();
        }
        /** @type {?} */
        var sub = this.eventSubject.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            eventHandler(value);
        }));
        this.observers.set(observerKey, sub);
        return this;
    };
    /**
     * @param {?} eventHandler
     * @return {?}
     */
    EventBusPipe.prototype.subscribeOnce = /**
     * @param {?} eventHandler
     * @return {?}
     */
    function (eventHandler) {
        /** @type {?} */
        var self = this;
        /** @type {?} */
        var scription = this.eventSubject.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            eventHandler(value);
            if (scription) {
                scription.unsubscribe();
            }
            self.parentEventPipeList = new Array();
        }));
        return this;
    };
    /**
     * @return {?}
     */
    EventBusPipe.prototype.unSubscribe = /**
     * @return {?}
     */
    function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
            this.subscription = null;
        }
    };
    // 注销使用once方法注册的订阅。
    // 注销使用once方法注册的订阅。
    /**
     * @return {?}
     */
    EventBusPipe.prototype.unSubscribeForOnce = 
    // 注销使用once方法注册的订阅。
    /**
     * @return {?}
     */
    function () {
        this.onceSubscription.unsubscribe();
    };
    /**
     * 取消某观察者的订阅
     * @param observerKey 观察者唯一标识
     */
    /**
     * 取消某观察者的订阅
     * @param {?} observerKey 观察者唯一标识
     * @return {?}
     */
    EventBusPipe.prototype.unSubscribeWithObkey = /**
     * 取消某观察者的订阅
     * @param {?} observerKey 观察者唯一标识
     * @return {?}
     */
    function (observerKey) {
        /** @type {?} */
        var scription = this.observers.get(observerKey);
        if (scription) {
            scription.unsubscribe();
        }
    };
    /**
     * @param {?} tokenValue
     * @return {?}
     */
    EventBusPipe.prototype.matchEmitterToken = /**
     * @param {?} tokenValue
     * @return {?}
     */
    function (tokenValue) {
        if (this.tokenValue && tokenValue && this.tokenValue !== tokenValue) {
            return false;
        }
        return true;
    };
    /**
     * @param {?} tokenValue
     * @return {?}
     */
    EventBusPipe.prototype.examByTargetToken = /**
     * @param {?} tokenValue
     * @return {?}
     */
    function (tokenValue) {
        if (this.tokenValue !== tokenValue) {
            return false;
        }
        return true;
    };
    /**
     * @param {?} subscriber
     * @return {?}
     */
    EventBusPipe.prototype.dispose = /**
     * @param {?} subscriber
     * @return {?}
     */
    function (subscriber) {
    };
    EventBusPipe.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    EventBusPipe.ctorParameters = function () { return [
        { type: String },
        { type: Array },
        { type: Subject }
    ]; };
    return EventBusPipe;
}());
export { EventBusPipe };
if (false) {
    /** @type {?} */
    EventBusPipe.prototype.eventSubject;
    /** @type {?} */
    EventBusPipe.prototype.observers;
    /** @type {?} */
    EventBusPipe.prototype.pipeCode;
    /**
     * @type {?}
     * @private
     */
    EventBusPipe.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    EventBusPipe.prototype.onceSubscription;
    /**
     * @type {?}
     * @private
     */
    EventBusPipe.prototype.tokenValue;
    /**
     * @type {?}
     * @private
     */
    EventBusPipe.prototype.parentEventPipeList;
}
