// @farris/mobile-bef/fesm2015/farris-mobile-bef.js
import { InjectionToken, HttpUtil, BindingPathConverter, PropertyUtil, BindingPropertyType, FieldMetadataUtil, ModifyType, EntityCollection, HttpMethods, makePropDecorator, EntityManager, DataChangeType, AppContext, MetadataUtil, InjectFlags, HttpClient, Repository } from '@farris/mobile-devkit';
import { BehaviorSubject, Subject, of, EMPTY, zip } from 'rxjs';
import { switchMap, tap, map, catchError } from 'rxjs/operators';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/types.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 变更类型
 */
class ChangeDetailType {
}
/**
 * 新增
 */
ChangeDetailType.Added = 'Added';
/**
 * 修改
 */
ChangeDetailType.Modify = 'Modify';
/**
 * 删除
 */
ChangeDetailType.Deleted = 'Deleted';
if (false) {
    /**
     * 新增
     * @type {?}
     */
    ChangeDetailType.Added;
    /**
     * 修改
     * @type {?}
     */
    ChangeDetailType.Modify;
    /**
     * 删除
     * @type {?}
     */
    ChangeDetailType.Deleted;
}
/**
 * 行变更信息
 * 必须包含：
 * 1、DataId   => 主键值；
 * 2、属性名   => 新的属性值；
 * 3、子表名+s => 子表行的ChangeDetail数组
 * @record
 */
function ChangeDetailInfo() { }
if (false) {
    /** @type {?|undefined} */
    ChangeDetailInfo.prototype.DataId;
    /* Skipping unhandled member: [key: string]: number | string | boolean | null | ChangeDetail | ChangeDetail[] | { [key: string]: any };*/
}
/**
 * 行变更详情
 * 包含：
 * 1、变更类型；
 * 2、变更信息
 * @record
 */
function ChangeDetail() { }
if (false) {
    /** @type {?} */
    ChangeDetail.prototype.ChangeType;
    /** @type {?} */
    ChangeDetail.prototype.ChangeInfo;
}
/**
 * 请求类型
 * @record
 */
function RequestInfo() { }
if (false) {
    /** @type {?} */
    RequestInfo.prototype.dataChange;
    /** @type {?|undefined} */
    RequestInfo.prototype.variableChange;
}
/**
 * 分页信息
 * @record
 */
function Pagination() { }
if (false) {
    /** @type {?} */
    Pagination.prototype.pageSize;
    /** @type {?} */
    Pagination.prototype.totalCount;
    /** @type {?} */
    Pagination.prototype.pageCount;
    /** @type {?} */
    Pagination.prototype.pageIndex;
}
/**
 * 查询结果
 * @record
 */
function QueryResult() { }
if (false) {
    /** @type {?} */
    QueryResult.prototype.result;
    /** @type {?} */
    QueryResult.prototype.pagination;
}
/**
 * 返回结果类型
 * @record
 */
function ResponseInfo() { }
if (false) {
    /** @type {?} */
    ResponseInfo.prototype.returnValue;
    /** @type {?} */
    ResponseInfo.prototype.message;
    /** @type {?} */
    ResponseInfo.prototype.innerDataChange;
    /** @type {?} */
    ResponseInfo.prototype.innerVariableChange;
}
/**
 * 带RequestInfo的body对象
 * @record
 */
function BodyWithRequestInfo() { }
if (false) {
    /** @type {?} */
    BodyWithRequestInfo.prototype.requestInfo;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * request option格式
 * @record
 */
function RequestOption() { }
if (false) {
    /** @type {?} */
    RequestOption.prototype.body;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 *  异常处理信息接口
 * @record
 */
function IErrorServe() { }
if (false) {
    /**
     * @param {?} error
     * @return {?}
     */
    IErrorServe.prototype.show = function (error) { };
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/tokens.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const BE_SERVER_URI_TOKEN = new InjectionToken('@farris/be BE_SERVER_URL');
/**
 * 1、因为bef里，很多类都没有注入，BefSessionHandlingStragegy无法直接注入；
 * 2、通过一个字符串Token来间接做策略选择。
 * @type {?}
 */
const BE_SESSION_HANDLING_STRATEGY_TOKEN = new InjectionToken('@farris/be BE_SESSION_HANDLING_STRATEGY_TOKEN');
/**
 * 注入异常处理service
 * @type {?}
 */
const BE_ERROR_HANDLER__TOKEN = new InjectionToken('@farris/be BE_ERROR_HANDLER__TOKEN');

/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/bef_http.util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * BefHttp工具类
 */
class BefHttpUtil {
    /**
     * 追加SessionId头
     * @param {?} headers
     * @param {?} sessionId
     * @return {?}
     */
    static appendSessionId(headers, sessionId) {
        headers = HttpUtil.appendHeader(headers, 'SessionId', sessionId);
        return headers;
    }
    /**
     * 追加CommonVariable头
     * \@summary
     * 框架会话token，等价于原来的SessionId
     * @param {?} headers
     * @param {?} commonVariable
     * @return {?}
     */
    static appendCafRuntimeCommonVariable(headers, commonVariable) {
        headers = HttpUtil.appendHeader(headers, 'X-CAF-Runtime-CommonVariable', commonVariable);
        return headers;
    }
    /**
     * 追加X-CAF-Runtime-Context头
     * \@summary
     * X-CAF-Runtime-Context等价于BeSessionId
     * @param {?} headers
     * @param {?} context
     * @return {?}
     */
    static appendCafRuntimeContext(headers, context) {
        headers = HttpUtil.appendHeader(headers, 'X-CAF-Runtime-Context', context);
        return headers;
    }
    /**
     * 追加Content-Type头
     * \@summary
     * 提交内容的MIME类型，默认为application/json
     * @param {?} headers
     * @param {?=} contentType
     * @return {?}
     */
    static appendContextType(headers, contentType) {
        contentType = contentType ? contentType : 'application/json';
        headers = HttpUtil.appendHeader(headers, 'Content-Type', contentType);
        return headers;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/request_info.util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class RequestInfoUtil {
    /**
     * 构造RequestInfo
     * @param {?} repository
     * @return {?}
     */
    static buildRequestInfo(repository) {
        /** @type {?} */
        const requestInfo = {
            dataChange: repository.entityManager.buildAllEntityChangeDetails(),
            variableChange: repository.variableManager.buildChangeDetail()
        };
        return requestInfo;
    }
    /**
     * 向body中添加RequestInfo
     * @param {?} body
     * @param {?} repository
     * @return {?}
     */
    static appendRequestInfoToBody(body, repository) {
        if (body.requestInfo) {
            return body;
        }
        /** @type {?} */
        const requestInfo = this.buildRequestInfo(repository);
        // body不存在时，body=requestInfo
        if (!body || Object.keys(body).length === 0) {
            return requestInfo;
        }
        /** @type {?} */
        const bodyWithRequestInfo = Object.assign({}, body, { requestInfo: requestInfo });
        return bodyWithRequestInfo;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/response_info.util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ResponseInfoUtil {
    /**
     * 将ResponseInfo转换为老接口的数据返回格式
     * @param {?} result
     * @return {?}
     */
    static unWrapResponseInfo(result) {
        // 兼容cancel没有返回值的情况
        if (!result) {
            return result;
        }
        // 没有returnValue的情况下，兼容query取数的的格式
        if (result.hasOwnProperty('returnValue') === false) {
            if (result.hasOwnProperty('result') && result.hasOwnProperty('pagination')) {
                // 兼容返回带分页的列表数据
                return result.result;
            }
            return result;
        }
        // 其他：返回RequestInfo.returnValue的情况
        /** @type {?} */
        const returnValue = result.returnValue;
        if (returnValue && returnValue.hasOwnProperty('result') && returnValue.hasOwnProperty('pagination')) {
            // 兼容返回带分页的列表数据
            return returnValue.result;
        }
        return result.returnValue;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/bef_change_util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 变更集工具类
 */
class BefChangeUtil {
    /**
     * 创建空的ChangeDetail
     * @param {?} type   the type of change
     * @param {?=} dataId the identity of the entity whose properties's values have changed
     * @return {?}
     */
    static createEmpty(type, dataId) {
        /** @type {?} */
        const changeDetail = {
            ChangeType: type,
            ChangeInfo: {}
        };
        if (dataId) {
            changeDetail.ChangeInfo.DataId = dataId;
        }
        return changeDetail;
    }
    /**
     * Get changes from an ChangeDetail object or an plain object
     * @param {?} changeDetail 变更详情
     * @return {?}
     */
    static getChangeInfo(changeDetail) {
        /** @type {?} */
        const isChangeDetail = this.isChangeDetail(changeDetail);
        if (isChangeDetail === true) {
            return changeDetail.ChangeInfo;
        }
        else {
            return changeDetail;
        }
    }
    /**
     * Check whether the changeDetial object is an instance of the ChangeDetial class.
     * @param {?} changeDetail
     * @return {?}
     */
    static isChangeDetail(changeDetail) {
        // @todo: if the associated object has only two properties: ChangeType and ChangeInfo, it fails.
        /** @type {?} */
        const keys = Object.keys(changeDetail);
        return keys.length === 2 && keys.indexOf('ChangeType') > -1 && keys.indexOf('ChangeInfo') > -1;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/bef_data_path_util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Bef数据中各种Path的转换
 */
class BefDataPathUtil {
    /**
     * 转换成路径数组
     * 返回结果：
     * 主表（/）：[ parentId ]
     * 从表（/childCodes）：[ parentId, childCodes, childId ]
     * 从从表（/childCodes/grandCodes）： [ parentId, childCode, childId, grandCode, grandId ]
     * @param {?} bindingPath
     * @param {?} bindingData
     * @return {?}
     */
    static convertToPathArray(bindingPath, bindingData) {
        /** @type {?} */
        const bindingPathArray = BindingPathConverter.toBindingPathArray(bindingPath);
        /** @type {?} */
        const pathArray = [];
        /** @type {?} */
        let currentBindingObject = bindingData.list.currentItem;
        pathArray.push(currentBindingObject.primaryKeyValue);
        bindingPathArray.forEach((/**
         * @param {?} propName
         * @return {?}
         */
        (propName) => {
            /** @type {?} */
            const propInfo = PropertyUtil.getPropertyByName(currentBindingObject.properties, propName);
            if (propInfo.type !== BindingPropertyType.List) {
                throw new Error(`${propInfo.name}不是子表对应的属性`);
            }
            /** @type {?} */
            const currentBindingList = currentBindingObject[propName];
            currentBindingObject = currentBindingList.currentItem;
            pathArray.push(this.trimTrailingS(propName));
            pathArray.push(currentBindingObject.primaryKeyValue);
        }));
        return pathArray;
    }
    /**
     * 转换为RestUrl里的路径
     *
     * 返回结果：
     * 主表（/）：/
     * 从表（/jiwtEdus）：/xxx/jiwtEdu
     * 从从表（/jiwtEdus/jiwtGrades）： /xxx/jiwtEdu/xxx/jiwtGrade
     * @param {?} bindingPath
     * @param {?} bindingData
     * @return {?}
     */
    static convertToPathUrl(bindingPath, bindingData) {
        /** @type {?} */
        const pathArray = this.convertToPathArray(bindingPath, bindingData);
        pathArray.pop();
        return '/' + pathArray.join('/');
    }
    /**
     * 转换为表名数组
     * 返回结果：
     * 主表（/）：[]
     * 从表（/childCodes）：[ childCodes ]
     * 从从表（/childCodes/grandCodes [childCode ,grandCode]
     * @param {?} bindingPath
     * @param {?} bindingData
     * @return {?}
     */
    static convertToObjectCodes(bindingPath, bindingData) {
        /** @type {?} */
        const pathArray = this.convertToPathArray(bindingPath, bindingData);
        /** @type {?} */
        const pathLen = pathArray.length;
        /** @type {?} */
        const objectCodes = [];
        for (let i = 1; i < pathLen; i = i + 2) {
            objectCodes.push(pathArray[i]);
        }
        return objectCodes;
    }
    /**
     * 转换为id数组，包含最后一级的主键
     * 主表（/）：[ parentId ]
     * 从表（/childCodes）：[ parentId, childId ]
     * 从从表（/childCodes/grandCodes）： [ parentId, childId, grandId]
     * @param {?} bindingPath
     * @param {?} bindingData
     * @return {?}
     */
    static convertToDataIdsForUpdate(bindingPath, bindingData) {
        /** @type {?} */
        const pathArray = this.convertToPathArray(bindingPath, bindingData);
        /** @type {?} */
        const pathLen = pathArray.length;
        /** @type {?} */
        const dataIds = [];
        for (let i = 0; i < pathLen; i = i + 2) {
            dataIds.push(pathArray[i]);
        }
        return dataIds;
    }
    /**
     * 转换为id数组，不包含最后一级的主键
     * 主表（/）：[ ]
     * 从表（/childCodes）：[ parentId ]
     * 从从表（/childCodes/grandCodes）： [ parentId, childId]
     * @param {?} bindingPath
     * @param {?} bindingData
     * @return {?}
     */
    static convertToDataIdsForAdd(bindingPath, bindingData) {
        /** @type {?} */
        const dataIds = this.convertToDataIdsForUpdate(bindingPath, bindingData);
        dataIds.pop();
        return dataIds;
    }
    /**
     * 将EntityPathString转换为Url
     * @deprecated
     * @param {?} path
     * @return {?}
     */
    static convertPathToUrl(path) {
        /** @type {?} */
        const subPaths = path.split('/');
        for (let i = subPaths.length - 1; i > 0; i--) {
            if (subPaths[i] && subPaths[i].endsWith('s')) {
                subPaths[i] = subPaths[i].substr(0, subPaths[i].length - 1);
            }
        }
        return subPaths.join('/').toLowerCase();
    }
    /**
     * 去除最后的s
     * @private
     * @param {?} str
     * @return {?}
     */
    static trimTrailingS(str) {
        return str.substr(0, str.length - 1);
    }
    /**
     * 转换通用路径为仅有表名的数组
     *
     * 1.['id:xxx','child1s','id:xx','prop'] => ['child1s','prop']
     * 2.['prop1','prop2','prop3'] => ['prop1','prop2','prop3']
     * 3.['prop1','id:xxx','prop2'] => ['prop1','prop2']
     * @param {?} paths 内置路径，格式:['id:xxx','child1s','id:xx','prop']
     * @return {?}
     */
    static convertPathsToNodeCodes(paths) {
        /** @type {?} */
        const result = [];
        if (!paths || paths.length < 1) {
            return result;
        }
        paths.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => !!item && item.indexOf(':') === -1)).forEach((/**
         * @param {?} node
         * @return {?}
         */
        node => {
            result.push(node);
        }));
        return result;
    }
    /**
     * 转换通用路径为仅有实体主键值的数组
     * @param {?} paths 通用路径，格式:['id:xxx','child1s','id:xx','prop'] => ['xxx','xx']
     * @return {?}
     */
    static convertPathsToIds(paths) {
        /** @type {?} */
        const result = [];
        if (!paths || paths.length < 1) {
            return result;
        }
        paths.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => !!item && item.indexOf(':') !== -1)).forEach((/**
         * @param {?} id
         * @return {?}
         */
        id => {
            result.push(id.split(':')[1]);
        }));
        return result;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/bef_proxy_util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * BefProxy工具类
 * \@summary
 * 1、BefProxy辅助工具类；
 * 2、暂时无法归类的代码封装；
 */
class BefProxyUtil {
    /**
     * 构造RequestInfo
     * @param {?} befRepository
     * @return {?}
     */
    static buildRequestInfo(befRepository) {
        /** @type {?} */
        const requestInfo = {
            dataChange: befRepository.entityManager.buildAllEntityChangeDetails(),
            variableChange: befRepository.variableManager.buildChangeDetail()
        };
        return requestInfo;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/bef_repository_util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class BefRepositoryUtil {
    /**
     * isExistUnsaveData
     * @param {?} befRepository
     * @return {?}
     */
    static isExistUnsaveData(befRepository) {
        /** @type {?} */
        let hasUnsavedData = false;
        if (!befRepository || !befRepository.entityCollection) {
            throw new Error('Current Object is null or it\'s entityCollection is null.');
        }
        /** @type {?} */
        const entityArray = befRepository.entityCollection.toArray();
        if (befRepository.dataChangeHistory.isChanged()) {
            return true;
        }
        for (let i = 0; i < entityArray.length; i++) {
            if (entityArray[i].changes.length > 0) {
                hasUnsavedData = true;
                break;
            }
        }
        return hasUnsavedData;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/entity_util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class EntityUtil {
    /**
     * 查找属性的类型
     * @param {?} entityType 实体类型
     * @param {?} targetPropName 属性名称
     * @return {?} 属性信息，包含属性类型（NgField、NgObject、NgList, NgDynamic）和属性对应的实体类型（当NgField类型时为null）
     */
    static getPropInfo(entityType, targetPropName) {
        /** @type {?} */
        let propType;
        /** @type {?} */
        let propEntityType;
        /** @type {?} */
        let propMetadata;
        // NgField
        /** @type {?} */
        const ngFieldProperties = FieldMetadataUtil.getNgFields(entityType);
        Object.keys(ngFieldProperties).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        (propName) => {
            if (propName === targetPropName) {
                propType = 'NgField';
                propEntityType = null;
                propMetadata = ngFieldProperties[propName];
            }
        }));
        // NgObject
        /** @type {?} */
        const ngObjectProperties = FieldMetadataUtil.getNgObjects(entityType);
        Object.keys(ngObjectProperties).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        (propName) => {
            if (propName === targetPropName) {
                propType = 'NgObject';
                propEntityType = ngObjectProperties[propName].type;
                propMetadata = ngObjectProperties[propName];
            }
        }));
        // NgList
        /** @type {?} */
        const ngListProperties = FieldMetadataUtil.getNgList(entityType);
        Object.keys(ngListProperties).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        (propName) => {
            if (propName === targetPropName) {
                propType = 'NgList';
                propEntityType = ngListProperties[propName].type;
                propMetadata = ngListProperties[propName];
            }
        }));
        /** @type {?} */
        const ngDynamicProperties = FieldMetadataUtil.getNgDynamic(entityType);
        Object.keys(ngDynamicProperties).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        (propName) => {
            if (propName === targetPropName) {
                propType = 'NgDynamic';
                propEntityType = ngDynamicProperties[propName].type;
                propMetadata = ngDynamicProperties[propName];
            }
        }));
        return { propType, propEntityType, propMetadata };
    }
    /**
     * 获取实体的主键名
     * @param {?} entityType 实体类型
     * @return {?}
     */
    static getPrimaryKey(entityType) {
        /** @type {?} */
        const primaryNgFiledProp = FieldMetadataUtil.getPrimaryFieldMetadata(entityType);
        if (primaryNgFiledProp) {
            return primaryNgFiledProp.dataField;
        }
        else {
            return '';
        }
    }
    /**
     * 是否为对象属性
     * @param {?} entityType
     * @param {?} targetPropName
     * @return {?}
     */
    static isObjectProp(entityType, targetPropName) {
        /** @type {?} */
        let isObjectProp = false;
        /** @type {?} */
        const ngObjectProperties = FieldMetadataUtil.getNgObjects(entityType);
        Object.keys(ngObjectProperties).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        (propName) => {
            if (propName === targetPropName) {
                isObjectProp = true;
            }
        }));
        return isObjectProp;
    }
    /**
     * 检查是否是动态列属性
     * @param {?} entityType
     * @param {?} targetPropName
     * @return {?}
     */
    static isDynamicProp(entityType, targetPropName) {
        /** @type {?} */
        let isDynamicProp = false;
        /** @type {?} */
        const ngDynamicProperties = FieldMetadataUtil.getNgDynamic(entityType);
        Object.keys(ngDynamicProperties).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        (propName) => {
            if (propName === targetPropName) {
                isDynamicProp = true;
            }
        }));
        return isDynamicProp;
    }
    /**
     * 为实体增加initialData属性
     * @param {?} entity 实体实例
     * @param {?} initialData 默认值对象
     * @return {?}
     */
    static appendInitialData(entity, initialData) {
        /** @type {?} */
        const data = Object.assign({}, initialData);
        delete data.id;
        delete data.parentID;
        entity['initialData'] = data;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/bef_env_util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Bef环境监测工具类
 */
class BefEnvUtil {
    /**
     * 是否在框架内运行
     * @return {?}
     */
    static isInFramework() {
        /** @type {?} */
        const hashString = window.location.hash;
        if (!hashString) {
            return false;
        }
        return hashString.indexOf('formToken=') !== -1;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_change_builder.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * BEF变更集构造器
 */
class BefChangeBuilder {
    /**
     * 构造函数
     * @param {?} entityType 实体类型
     * @param {?} entityCollection
     */
    constructor(entityType, entityCollection) {
        this.entityType = entityType;
        this.entityCollection = entityCollection;
    }
    /**
     * 构造Bef变更集
     * @param {?} modifications
     * @return {?}
     */
    build(modifications) {
        // 重置changeDetail
        this.changeDetail = {
            ChangeType: ChangeDetailType.Modify,
            ChangeInfo: {
                DataId: ''
            }
        };
        modifications.forEach((/**
         * @param {?} modification
         * @return {?}
         */
        modification => {
            this.buildChangeDetail(modification);
        }));
        return this.changeDetail;
    }
    /**
     * 构造Bef变更详情
     * @param {?} modification
     * @return {?}
     */
    buildChangeDetail(modification) {
        /** @type {?} */
        const paths = modification.path.concat();
        // 设置根节点DataId
        if (!this.changeDetail.ChangeInfo.DataId) {
            this.changeDetail.ChangeInfo.DataId = paths[0].split(':')[1];
        }
        /** @type {?} */
        let parentChangeDetail = this.changeDetail;
        /** @type {?} */
        let parentEntityType = this.entityType;
        for (let i = 1; i < paths.length && parentChangeDetail; i = i + 2) {
            /** @type {?} */
            const parentChangeInfo = this.getChangeInfo(parentChangeDetail);
            /** @type {?} */
            const propName = paths[i];
            const { propType, propEntityType, propMetadata } = EntityUtil.getPropInfo(parentEntityType, propName);
            /** @type {?} */
            const dataField = propMetadata.dataField || propName;
            if (propType === 'NgField') {
                // 不支持主键变更，忽略
                /** @type {?} */
                const primaryKey = EntityUtil.getPrimaryKey(parentEntityType);
                if (propName === primaryKey) {
                    continue;
                }
                if (modification.type !== ModifyType.ValueChange) {
                    throw Error('简单类型的属性上不支持ValueChange类型之外的变更');
                }
                // NgField类型：说明是最后一级
                parentChangeInfo[dataField] = modification.value;
                parentChangeDetail = null;
            }
            else if (propType === 'NgObject') {
                // NgObject属性本身无法触发变更，只有它的子节点才能触发，所以它上边的变更永远是Modify类型的。
                /** @type {?} */
                const childId = paths[i + 1].split(':')[1];
                /** @type {?} */
                const childIdName = paths[i + 1].split(':')[0];
                if (childIdName) {
                    // 有主键（关联对象）：是一个普通的对象
                    /** @type {?} */
                    let changeObject = parentChangeInfo[dataField];
                    // 获取数据
                    /** @type {?} */
                    const entityPath = paths.slice(0, i + 1);
                    /** @type {?} */
                    const changedEntity = this.entityCollection.getEntityByPath(entityPath);
                    changeObject = changedEntity ? changedEntity.toJSON() : {};
                    parentChangeInfo[dataField] = changeObject;
                    parentChangeDetail = null;
                    parentEntityType = null;
                }
                else {
                    // 没有主键（值对象）：是一个完整的ChangeDetail
                    /** @type {?} */
                    let changeDetail = (/** @type {?} */ (parentChangeInfo[dataField]));
                    if (!changeDetail) {
                        changeDetail = {
                            ChangeType: ChangeDetailType.Modify,
                            ChangeInfo: {}
                        };
                    }
                    parentChangeInfo[dataField] = changeDetail;
                    parentChangeDetail = changeDetail;
                    parentEntityType = propEntityType;
                }
            }
            else if (propType === 'NgList') {
                // 如果不存在则创建一个空数组
                if (!parentChangeDetail.ChangeInfo[dataField]) {
                    parentChangeDetail.ChangeInfo[dataField] = [];
                }
                /** @type {?} */
                const changeDetails = (/** @type {?} */ (parentChangeDetail.ChangeInfo[dataField]));
                // 如果这个属性，不是叶子节点，需要查找当前属性是否已经存在对应ChangeDetail：
                // 1、不存在：创建一个Modify类型的ChangeDetail；
                // 2、存在：返回查找到的ChangeDetai，这个ChangeDetail可能是一个Add类型也可能是一个Modify类型；
                // 3、现状：目前BEF不支持Add类型的变更，肯定是一个Modify类型的变更。
                if (i !== paths.length - 1) {
                    // 遍历检查变更是否已经存在
                    /** @type {?} */
                    const dataId = paths[i + 1].split(':')[1];
                    /** @type {?} */
                    let changeDetail = changeDetails.find((/**
                     * @param {?} changeDetailItem
                     * @return {?}
                     */
                    changeDetailItem => {
                        return changeDetailItem.ChangeInfo.DataId === dataId;
                    }));
                    // 如果不存在，则创建并添加
                    if (!changeDetail) {
                        changeDetail = this.createEmptyChangeDetail(ChangeDetailType.Modify, dataId);
                        changeDetails.push(changeDetail);
                    }
                    parentChangeDetail = changeDetail;
                    parentEntityType = propEntityType;
                    continue;
                }
                // 重置
                parentChangeDetail = null;
                parentEntityType = null;
            }
            else if (propType === 'NgDynamic') {
                // 获取数据
                /** @type {?} */
                const entityPath = paths.slice(0, i + 1);
                /** @type {?} */
                const changedEntity = this.entityCollection.getEntityByPath(entityPath);
                parentChangeInfo[dataField] = {
                    ChangeType: ChangeDetailType.Modify,
                    ChangeInfo: changedEntity ? changedEntity.toJSON() : {}
                };
                parentChangeDetail = null;
                parentEntityType = null;
            }
        }
    }
    /**
     * 获取变更信息
     * 在整个ChangeDetail树上，存在两种类型的节点
     * ChangeDetail：实体变更、值对象变更（没有DataID）
     * PlainObject: 关联对象的变更
     * 从这两种节点上拿具体变更信息的时候，需要统一处理，屏蔽这个差异。
     * \@todo：为这两种节点封装ChangeNode基类来解决这个差异。
     * @private
     * @param {?} changeDetail
     * @return {?}
     */
    getChangeInfo(changeDetail) {
        // @todo：可能存在同名属性
        if (changeDetail.hasOwnProperty('ChangeInfo')) {
            return changeDetail.ChangeInfo;
        }
        else {
            return changeDetail;
        }
    }
    /**
     * 创建ChangeDetail
     * @private
     * @param {?} type BEF变更类型
     * @param {?} dataId 数据内码
     * @return {?}
     */
    createEmptyChangeDetail(type, dataId) {
        /** @type {?} */
        const changeDetail = {
            ChangeType: type,
            ChangeInfo: {
                DataId: dataId
            }
        };
        return changeDetail;
    }
}
if (false) {
    /**
     * Bef变更集
     * @type {?}
     */
    BefChangeBuilder.prototype.changeDetail;
    /**
     * @type {?}
     * @private
     */
    BefChangeBuilder.prototype.entityType;
    /**
     * @type {?}
     * @private
     */
    BefChangeBuilder.prototype.entityCollection;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_change_handler.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 处理服务器端变更
 */
class BefChangeHandler {
    /**
     * 构造函数
     */
    constructor() {
    }
    /**
     * 处理Bef变更集
     * @param {?} entityType
     * @param {?} entityCollection
     * @param {?} changeDetails
     * @return {?}
     */
    handle(entityType, entityCollection, changeDetails) {
        this.handleChangeDetails(entityType, entityCollection, changeDetails);
    }
    /**
     * 处理Bef变更集（批量）
     * @param {?} entityType
     * @param {?} entityList
     * @param {?} changeDetails
     * @return {?}
     */
    handleChangeDetails(entityType, entityList, changeDetails) {
        if (!changeDetails) {
            return;
        }
        changeDetails.forEach((/**
         * @param {?} changeDetail
         * @return {?}
         */
        (changeDetail) => {
            /** @type {?} */
            const id = (/** @type {?} */ ((changeDetail.ChangeInfo.dataId || changeDetail.ChangeInfo.DataId)));
            /** @type {?} */
            const entity = this.getEntityById(entityList, id);
            if (!entity) {
                return;
            }
            this.handleChangeDetail(entityType, entity, changeDetail);
        }));
    }
    /**
     * 处理Bef变更集（单条）
     * @param {?} entityType
     * @param {?} entity
     * @param {?} changeDetail
     * @return {?}
     */
    handleChangeDetail(entityType, entity, changeDetail) {
        if (!changeDetail || !entity) {
            return;
        }
        // 只处理值变更，其他变更待进一步验证。
        if (changeDetail.ChangeType !== ChangeDetailType.Modify) {
            return;
        }
        /** @type {?} */
        const changeInfo = changeDetail.ChangeInfo;
        Object.keys(changeInfo).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        (propName) => {
            const { propType, propEntityType } = EntityUtil.getPropInfo(entityType, propName);
            if (propType === 'NgField') {
                // 简单属性：更新值
                entity[propName] = changeInfo[propName];
            }
            else if (propType === 'NgObject') {
                /** @type {?} */
                const childEntity = (/** @type {?} */ (entity[propName]));
                if (childEntity.primaryKey) {
                    // 关联对象：重新加载数据
                    /** @type {?} */
                    const childEntityData = changeInfo[propName];
                    childEntity.load(childEntityData);
                }
                else {
                    // 值对象：递归处理变更
                    /** @type {?} */
                    const childChangeDetail = (/** @type {?} */ (changeInfo[propName]));
                    this.handleChangeDetail(propEntityType, childEntity, childChangeDetail);
                }
            }
            else if (propType === 'NgList') {
                // 子列表：递归处理变更集合
                /** @type {?} */
                const childEntityList = (/** @type {?} */ (entity[propName]));
                /** @type {?} */
                const childChangeDetails = (/** @type {?} */ (changeInfo[propName]));
                this.handleChangeDetails(propEntityType, childEntityList, childChangeDetails);
            }
        }));
    }
    /**
     * 根据id获取实体，屏蔽EntityCollection和EntityList之间的差异
     * @private
     * @param {?} entityList
     * @param {?} id
     * @return {?}
     */
    getEntityById(entityList, id) {
        /** @type {?} */
        let target;
        if (entityList instanceof EntityCollection) {
            target = entityList.getEntityById(id);
        }
        else {
            target = entityList.get(id);
        }
        return target ? target : null;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/session/handling-strategies/handling_strategy.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * BefSession处理策略类
 * @abstract
 */
class BefSessionHandlingStrategy {
    /**
     * 构造函数
     * @param {?} storageStrategy
     * @param {?} frmSessionService
     * @param {?} httpClient
     * @param {?} beBaseUrl
     */
    constructor(storageStrategy, frmSessionService, httpClient, beBaseUrl) {
        /**
         * BeSession是否存在
         */
        this.beSessionExisted = false;
        this.storageStrategy = storageStrategy;
        this.frmSessionService = frmSessionService;
        this.httpClient = httpClient;
        this.beBaseUrl = beBaseUrl;
        this.beCreateSessionUrl = `${beBaseUrl}/service/createsession`;
        this.beCloseSessionUrl = `${beBaseUrl}/service/closesession`;
        this.beSessionExisted$ = new BehaviorSubject(this.beSessionExisted);
    }
    /**
     * 框架SessionId（用户的或者功能菜单的）
     * @protected
     * @return {?}
     */
    get frmSessionId() {
        return this.frmSessionService.getCurrentSessionId();
    }
    /**
     * 获取框架SessionId
     * @return {?}
     */
    getFrameworkSessionId() {
        return this.frmSessionId;
    }
    /**
     * 从缓存中获取BeSession
     * @protected
     * @return {?}
     */
    getSessionIdFromStorage() {
        /** @type {?} */
        const sessionStorageKey = this.getSessionStorageKey();
        /** @type {?} */
        const beSessionId = this.storageStrategy.getItem(sessionStorageKey);
        return beSessionId;
    }
    /**
     * 创建BeSessionId
     * @protected
     * @return {?}
     */
    createSession() {
        /** @type {?} */
        const requestConfig = {
            responseType: 'text',
            headers: {
                'Content-Type': 'application/json'
            }
        };
        /** @type {?} */
        const frmSessionId = this.frmSessionId;
        if (frmSessionId) {
            requestConfig.headers = BefHttpUtil.appendSessionId(requestConfig.headers, this.frmSessionId);
            requestConfig.headers = BefHttpUtil.appendCafRuntimeCommonVariable(requestConfig.headers, this.frmSessionId);
        }
        //创建前判断当前的frmsessionid是不是在当前会话下被清除过
        /** @type {?} */
        const ClearItOnce = window.sessionStorage.getItem("ClearItOnce");
        if ('true' !== ClearItOnce && frmSessionId) {
            return this.clear(frmSessionId).pipe(switchMap((/**
             * @return {?}
             */
            () => {
                window.sessionStorage.setItem("ClearItOnce", 'true');
                return this.httpClient.post(this.beCreateSessionUrl, null, requestConfig).pipe(tap((/**
                 * @param {?} beSessionId
                 * @return {?}
                 */
                (beSessionId) => {
                    this.setSessionId(beSessionId);
                    this.setBesessionExisted(true);
                })));
            })));
        }
        return this.httpClient.post(this.beCreateSessionUrl, null, requestConfig).pipe(tap((/**
         * @param {?} beSessionId
         * @return {?}
         */
        (beSessionId) => {
            this.setSessionId(beSessionId);
            this.setBesessionExisted(true);
        })));
    }
    /**
     * @param {?} formToken
     * @return {?}
     */
    clear(formToken) {
        /** @type {?} */
        const subject = new Subject();
        if (formToken && window['frmMobileService']) {
            window['frmMobileService'].rtf.func.clearState({ formToken }, (/**
             * @return {?}
             */
            () => { subject.next(true); }));
        }
        return subject;
    }
    /**
     * @return {?}
     */
    test() {
        /** @type {?} */
        const requestConfig = {
            responseType: 'text',
            headers: {
                'Content-Type': 'application/json'
            }
        };
        if (this.frmSessionId) {
            requestConfig.headers = BefHttpUtil.appendSessionId(requestConfig.headers, this.frmSessionId);
            requestConfig.headers = BefHttpUtil.appendCafRuntimeCommonVariable(requestConfig.headers, this.frmSessionId);
        }
        return this.httpClient.post(this.beCreateSessionUrl, null, requestConfig).pipe(tap((/**
         * @param {?} beSessionId
         * @return {?}
         */
        (beSessionId) => {
            this.setSessionId(beSessionId);
            this.setBesessionExisted(true);
        })));
    }
    /**
     * 关闭BeSessionId
     * @protected
     * @return {?}
     */
    closeOldSession() {
        if (!this.oldBeSessionId) {
            return of(true);
        }
        /** @type {?} */
        const requestConfig = {
            responseType: 'text',
            headers: {
                'Content-Type': 'application/json'
            }
        };
        // headers处理
        requestConfig.headers = BefHttpUtil.appendCafRuntimeContext(requestConfig.headers, this.oldBeSessionId);
        requestConfig.headers = BefHttpUtil.appendSessionId(requestConfig.headers, this.oldBeSessionId);
        if (this.frmSessionId) {
            requestConfig.headers = BefHttpUtil.appendCafRuntimeCommonVariable(requestConfig.headers, this.frmSessionId);
        }
        // 无论是否成功，统一置空cleardBeSessionId
        return this.httpClient.post(this.beCloseSessionUrl, null, requestConfig).pipe(tap((/**
         * @return {?}
         */
        () => {
            this.oldBeSessionId = null;
            return of(true);
        }), (/**
         * @return {?}
         */
        () => {
            this.oldBeSessionId = null;
            return of(true);
        })));
    }
    /**
     * 设置BeSession存在状态，并发出通知
     * @protected
     * @param {?} beSessionExisted
     * @return {?}
     */
    setBesessionExisted(beSessionExisted) {
        if (this.beSessionExisted === beSessionExisted) {
            return;
        }
        this.beSessionExisted = beSessionExisted;
        this.beSessionExisted$.next(beSessionExisted);
    }
    /**
     * 获取BeSession是否存在状态
     * @return {?}
     */
    getBeSessionExisted() {
        return this.beSessionExisted$;
    }
}
if (false) {
    /**
     * 存储策略
     * @type {?}
     * @protected
     */
    BefSessionHandlingStrategy.prototype.storageStrategy;
    /**
     * 框架Session服务
     * @type {?}
     * @protected
     */
    BefSessionHandlingStrategy.prototype.frmSessionService;
    /**
     * Http客户端
     * @type {?}
     * @protected
     */
    BefSessionHandlingStrategy.prototype.httpClient;
    /**
     * 创建Session的的EAPI地址
     * @type {?}
     * @protected
     */
    BefSessionHandlingStrategy.prototype.beBaseUrl;
    /**
     * 创建BeSession接口地址
     * @type {?}
     * @protected
     */
    BefSessionHandlingStrategy.prototype.beCreateSessionUrl;
    /**
     * 关闭BeSession接口地址
     * @type {?}
     * @protected
     */
    BefSessionHandlingStrategy.prototype.beCloseSessionUrl;
    /**
     * 清空的BeSessionId
     * @type {?}
     * @protected
     */
    BefSessionHandlingStrategy.prototype.oldBeSessionId;
    /**
     * BeSession是否存在
     * @type {?}
     * @private
     */
    BefSessionHandlingStrategy.prototype.beSessionExisted;
    /**
     * BeSession是否存在流
     * @type {?}
     * @private
     */
    BefSessionHandlingStrategy.prototype.beSessionExisted$;
    /**
     * @abstract
     * @return {?}
     */
    BefSessionHandlingStrategy.prototype.getSessionId = function () { };
    /**
     * @abstract
     * @param {?} sessionId
     * @return {?}
     */
    BefSessionHandlingStrategy.prototype.setSessionId = function (sessionId) { };
    /**
     * @abstract
     * @return {?}
     */
    BefSessionHandlingStrategy.prototype.clearSessionId = function () { };
    /**
     * @abstract
     * @param {?} headers
     * @return {?}
     */
    BefSessionHandlingStrategy.prototype.handleRequestHeaders = function (headers) { };
    /**
     * @abstract
     * @param {?} headers
     * @return {?}
     */
    BefSessionHandlingStrategy.prototype.handleReponseHeaders = function (headers) { };
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    BefSessionHandlingStrategy.prototype.getSessionStorageKey = function () { };
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/session/handling-strategies/separated_handling_strategy.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 隔离的BeSession处理策略
 */
class BefSeparatedSessionHandlingStrategy extends BefSessionHandlingStrategy {
    /**
     * 构造函数
     * @param {?} storageStrategy
     * @param {?} frmSessionService
     * @param {?} httpClient
     * @param {?} beBaseUrl
     */
    constructor(storageStrategy, frmSessionService, httpClient, beBaseUrl) {
        super(storageStrategy, frmSessionService, httpClient, beBaseUrl);
    }
    /**
     * 获取BeSessionId
     * @return {?}
     */
    getSessionId() {
        /** @type {?} */
        const beSessionId = this.getSessionIdFromStorage();
        /** @type {?} */
        let beSessionId$;
        if (beSessionId) {
            beSessionId$ = of(beSessionId);
            this.setBesessionExisted(true);
        }
        else {
            beSessionId$ = this.createSession();
        }
        /** @type {?} */
        const result$ = this.closeOldSession().pipe(switchMap((/**
         * @return {?}
         */
        () => {
            return beSessionId$;
        })));
        return result$;
    }
    /**
     * 设置BeSessionId
     * @param {?} sessionId
     * @return {?}
     */
    setSessionId(sessionId) {
        /** @type {?} */
        const sessionKey = this.getSessionStorageKey();
        this.storageStrategy.setItem(sessionKey, sessionId);
    }
    /**
     * 清空BeSessionId
     * @return {?}
     */
    clearSessionId() {
        if (BefEnvUtil.isInFramework() === true) {
            this.storageStrategy.removeItemsByScope(this.frmSessionId);
        }
        else {
            /** @type {?} */
            const sessionKey = this.getSessionStorageKey();
            this.oldBeSessionId = this.getSessionIdFromStorage();
            this.storageStrategy.removeItem(sessionKey);
        }
    }
    /**
     * 扩展BeSessionId相关头信息
     * @param {?} headers
     * @return {?}
     */
    handleRequestHeaders(headers) {
        /** @type {?} */
        const frmSessionId = this.getFrameworkSessionId();
        /** @type {?} */
        const beSessionId = this.getSessionIdFromStorage();
        if (frmSessionId) {
            headers = BefHttpUtil.appendCafRuntimeCommonVariable(headers, frmSessionId);
        }
        if (beSessionId) {
            headers = BefHttpUtil.appendCafRuntimeContext(headers, beSessionId);
            headers = BefHttpUtil.appendSessionId(headers, beSessionId);
        }
        return headers;
    }
    /**
     * 处理服务器端返回的headers
     * @param {?} headers
     * @return {?}
     */
    handleReponseHeaders(headers) {
    }
    /**
     * 获取某个Repository对应的BeSession的唯一key
     * @protected
     * @return {?}
     */
    getSessionStorageKey() {
        return `${this.frmSessionId}_${this.beBaseUrl}`;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/session/handling-strategies/unified_handling_strategy.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 统一的Session处理策略
 * \@todo
 * 1、目前Session其实还是隔离的，因为使用了beBaseUrl做了缓存；
 * 2、将来框架token问题解决了之后，使用token或者表单基url做缓存key；
 * 3、基于2缓存之后，如何closeSession，成为问题，除非appAcontext是一颗树，只有根AppContext初始化的时候，才去closeSession。
 */
class BefUnifiedSessionHandlingStrategy extends BefSessionHandlingStrategy {
    /**
     * 构造函数
     * @param {?} storageStrategy
     * @param {?} frmSessionService
     * @param {?} httpClient
     * @param {?} beBaseUrl
     */
    constructor(storageStrategy, frmSessionService, httpClient, beBaseUrl) {
        super(storageStrategy, frmSessionService, httpClient, beBaseUrl);
        this.beCloseSessionUrl = `${beBaseUrl}/service/closesession`;
    }
    /**
     * 获取BeSession
     * @return {?}
     */
    getSessionId() {
        /** @type {?} */
        const sessionKey = this.getSessionStorageKey();
        /** @type {?} */
        const sessionId = this.storageStrategy.getItem(sessionKey);
        /** @type {?} */
        const beSessionId$ = of(sessionId);
        /** @type {?} */
        const result$ = this.closeOldSession().pipe(switchMap((/**
         * @return {?}
         */
        () => {
            return beSessionId$;
        })));
        return result$;
    }
    /**
     * 设置BeSessionId
     * @param {?} sessionId
     * @return {?}
     */
    setSessionId(sessionId) {
        /** @type {?} */
        const sessionKey = this.getSessionStorageKey();
        this.storageStrategy.setItem(sessionKey, sessionId);
    }
    /**
     * 清空Sessionid
     * @return {?}
     */
    clearSessionId() {
        /** @type {?} */
        const sessionKey = this.getSessionStorageKey();
        this.oldBeSessionId = this.getSessionIdFromStorage();
        this.storageStrategy.removeItem(sessionKey);
    }
    /**
     * 扩展Session相关头信息
     * @param {?} headers
     * @return {?}
     */
    handleRequestHeaders(headers) {
        /** @type {?} */
        const frmSessionId = this.getFrameworkSessionId();
        if (frmSessionId) {
            headers = BefHttpUtil.appendCafRuntimeCommonVariable(headers, frmSessionId);
        }
        /** @type {?} */
        const beSessionId = this.getSessionIdFromStorage();
        if (beSessionId) {
            headers = BefHttpUtil.appendCafRuntimeContext(headers, beSessionId);
        }
        return headers;
    }
    /**
     * 处理服务器端返回的headers
     * @param {?} headers
     * @return {?}
     */
    handleReponseHeaders(headers) {
        /** @type {?} */
        const beSessionId = headers['befsessionid'];
        /** @type {?} */
        const oldBeSessionId = this.getSessionId();
        if (beSessionId && beSessionId !== oldBeSessionId) {
            this.setSessionId(beSessionId);
        }
    }
    /**
     * 获取某个Repository对应的BeSession的唯一key
     * @protected
     * @return {?}
     */
    getSessionStorageKey() {
        return this.beBaseUrl;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/session/storage-strategies/storage_strategy.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * BeSession缓存
 * @record
 */
function BeSessionStorageStrategy() { }
if (false) {
    /**
     * @param {?} beSessionKey
     * @return {?}
     */
    BeSessionStorageStrategy.prototype.getItem = function (beSessionKey) { };
    /**
     * @param {?} beSessionKey
     * @param {?} beSessionid
     * @return {?}
     */
    BeSessionStorageStrategy.prototype.setItem = function (beSessionKey, beSessionid) { };
    /**
     * @param {?} beSessionKey
     * @return {?}
     */
    BeSessionStorageStrategy.prototype.removeItem = function (beSessionKey) { };
    /**
     * @param {?} beSessionScope
     * @return {?}
     */
    BeSessionStorageStrategy.prototype.removeItemsByScope = function (beSessionScope) { };
    /**
     * @return {?}
     */
    BeSessionStorageStrategy.prototype.clear = function () { };
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/session/storage-strategies/session_storage_strategy.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 基于浏览器SessionStorage的BeSession缓存
 * \@summary
 * 1、SeparatedSession模式下：
 *  {
 *    BE_SESSION_ID: {
 *      frmSessionId1_beSessionUri1: beSessionId-1,
 *      frmSessionId2_beSessionUri2: beSessionId-2,
 *    }
 *  }
 * 2、UnifiedSession模式下（在Debug状态，同模式1）：
 * {
 *    BE_SESSION_ID: {
 *      frmSessionId1: beSessionId-1,
 *      frmSessionId2: beSessionId-2,
 *    }
 *  }
 */
class SessionStorageBeSessionStorageStrategy {
    constructor() {
        /**
         * 缓存Token
         */
        this.sessionStorageKey = 'BE_SESSION_ID';
    }
    /**
     * 获取值
     * @param {?} beSessionKey
     * @return {?}
     */
    getItem(beSessionKey) {
        /** @type {?} */
        const beSessions = this.getAllBeSessions();
        return beSessions[beSessionKey];
    }
    /**
     * 设置值
     * @param {?} beSessionKey
     * @param {?} beSessionId
     * @return {?}
     */
    setItem(beSessionKey, beSessionId) {
        /** @type {?} */
        const beSessions = this.getAllBeSessions();
        beSessions[beSessionKey] = beSessionId;
        this.setAllBeSessions(beSessions);
    }
    /**
     * 删除值
     * @param {?} beSessionKey
     * @return {?}
     */
    removeItem(beSessionKey) {
        /** @type {?} */
        const beSessions = this.getAllBeSessions();
        if (beSessions[beSessionKey]) {
            delete beSessions[beSessionKey];
        }
        this.setAllBeSessions(beSessions);
    }
    /**
     * 根据scope删除值
     * @param {?} beSessionScope
     * @return {?}
     */
    removeItemsByScope(beSessionScope) {
        /** @type {?} */
        const beSessions = this.getAllBeSessions();
        Object.keys(beSessions).forEach((/**
         * @param {?} beSessionKey
         * @return {?}
         */
        (beSessionKey) => {
            if (beSessionKey.startsWith(beSessionScope) === true) {
                delete beSessions[beSessionKey];
            }
        }));
        this.setAllBeSessions(beSessions);
    }
    /**
     * 清空所有会话
     * @return {?}
     */
    clear() {
        window.sessionStorage.removeItem(this.sessionStorageKey);
    }
    /**
     * 从SessionStorage中获取全部BeSessions
     * @private
     * @return {?}
     */
    getAllBeSessions() {
        /** @type {?} */
        const beSessionsJson = window.sessionStorage.getItem(this.sessionStorageKey);
        if (!beSessionsJson) {
            return {};
        }
        return JSON.parse(beSessionsJson);
    }
    /**
     * 设置全部BeSessions到SessionStorage
     * @param {?} beSessions
     * @return {?}
     */
    setAllBeSessions(beSessions) {
        /** @type {?} */
        const beSessionsString = JSON.stringify(beSessions);
        window.sessionStorage.setItem(this.sessionStorageKey, beSessionsString);
    }
}
if (false) {
    /**
     * 缓存Token
     * @type {?}
     * @private
     */
    SessionStorageBeSessionStorageStrategy.prototype.sessionStorageKey;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/session/storage-strategies/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/session/handling-strategies/handling_strategy_factory.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * BeSession处理策略工厂
 */
class BefSessionHandlingStrategyFactory {
    /**
     * 创建BeSession处理策略
     * @param {?} handlingStrategyName
     * @param {?} frmSessionService
     * @param {?} beBaseUrl
     * @param {?} httpClient
     * @return {?}
     */
    create(handlingStrategyName, frmSessionService, beBaseUrl, httpClient) {
        /** @type {?} */
        const storageStrategy = this.createStorageStrategy();
        if (handlingStrategyName === 'UnifiedSession') {
            return new BefUnifiedSessionHandlingStrategy(storageStrategy, frmSessionService, httpClient, beBaseUrl);
        }
        else {
            return new BefSeparatedSessionHandlingStrategy(storageStrategy, frmSessionService, httpClient, beBaseUrl);
        }
    }
    /**
     * 创建BeSession缓存策略
     * @private
     * @return {?}
     */
    createStorageStrategy() {
        return new SessionStorageBeSessionStorageStrategy();
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/session/handling-strategies/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/session/bef_session_service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * BefSessionService
 */
class BefSessionService {
    /**
     * 构造函数
     * @param {?} handlingStrategy
     */
    constructor(handlingStrategy) {
        this.handlingStrategy = handlingStrategy;
    }
    /**
     * 获取token
     * @return {?}
     */
    get token() {
        return this.handlingStrategy.getFrameworkSessionId();
    }
    /**
     * 获取BeSessionId
     * @return {?}
     */
    getBeSessionId() {
        return this.handlingStrategy.getSessionId();
    }
    /**
     * 获取BeSession是否存在状态
     * @return {?}
     */
    getBeSessionExisted() {
        return this.handlingStrategy.getBeSessionExisted();
    }
    /**
     * 设置sessionId
     * @param {?} sessionId sessionId
     * @return {?}
     */
    setBeSessionId(sessionId) {
        this.handlingStrategy.setSessionId(sessionId);
    }
    /**
     * 清空BeSessionId
     * @return {?}
     */
    clearBeSessionId() {
        this.handlingStrategy.clearSessionId();
    }
    /**
     * 扩展请求header
     * @param {?} headers
     * @return {?}
     */
    extendRequestHeaders(headers) {
        return this.handlingStrategy.handleRequestHeaders(headers);
    }
    /**
     * 处理响应header
     * @param {?} headers
     * @return {?}
     */
    handleResponseHeaders(headers) {
        return this.handlingStrategy.handleReponseHeaders(headers);
    }
}
if (false) {
    /**
     * Session处理策略类
     * @type {?}
     * @private
     */
    BefSessionService.prototype.handlingStrategy;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/session/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_proxy.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Bef API代理类
 * @abstract
 */
class BefProxy {
    /**
     * 构造函数
     * @param {?} httpClient
     */
    constructor(httpClient) {
        this.httpClient = httpClient;
        this.associatedUrlMap = new Map();
    }
    // 记录附件取跨Be url
    /**
     * @param {?} url
     * @return {?}
     */
    setAssociatedUrl(url) {
        this.associatedUrlMap.set(url, url);
    }
    /**
     * 设置扩展策略
     * @param {?} proxyExtend
     * @return {?}
     */
    setProxyExtend(proxyExtend) {
        this.proxyExtend = proxyExtend;
    }
    /**
     * 查询
     * @param {?=} entityFilter
     * @return {?}
     */
    query(entityFilter) {
        /** @type {?} */
        const url = this.baseUrl;
        /** @type {?} */
        const params = {};
        if (entityFilter) {
            /** @type {?} */
            const entityFilterString = JSON.stringify(entityFilter);
            params.entityFilter = entityFilterString;
        }
        /** @type {?} */
        const requestConfig = {
            params: params
        };
        return this.request(HttpMethods.GET, url, requestConfig);
    }
    /**
     * 列表数据查询（扩展）
     * @param {?} entityFilter
     * @return {?}
     */
    extendQuery(entityFilter) {
        /** @type {?} */
        const url = `${this.baseUrl}/extension/query`;
        /** @type {?} */
        const params = {};
        if (entityFilter) {
            /** @type {?} */
            const entityFilterString = JSON.stringify(entityFilter);
            params.entityFilter = entityFilterString;
        }
        /** @type {?} */
        const requestConfig = {
            params: params
        };
        return this.request(HttpMethods.PUT, url, requestConfig);
    }
    /**
     * 单条数据检索
     * @param {?} id
     * @return {?}
     */
    retrieve(id) {
        /** @type {?} */
        const url = `${this.baseUrl}/${id}`;
        return this.request(HttpMethods.GET, url);
    }
    /**
     * 单条数据检索（扩展）
     * @param {?} id
     * @return {?}
     */
    extendRetrieve(id) {
        /** @type {?} */
        const url = `${this.baseUrl}/extension/retrieve/${id}`;
        return this.request(HttpMethods.PUT, url);
    }
    /**
     * 单条数据检索（加锁）
     * @param {?} id
     * @return {?}
     */
    edit(id) {
        /** @type {?} */
        const url = `${this.baseUrl}/service/edit/${id}`;
        return this.request(HttpMethods.PUT, url);
    }
    /**
     * 新增数据
     * @param {?=} defaultValue
     * @param {?=} requestInfo
     * @return {?}
     */
    create(defaultValue, requestInfo) {
        /** @type {?} */
        const body = {
            defaultValue,
            requestInfo: requestInfo,
        };
        /** @type {?} */
        const requestConfig = {
            body: body
        };
        return this.request(HttpMethods.POST, this.baseUrl, requestConfig);
    }
    /**
     * 新增从表数据
     * \@path 新增路径（从表形如：/1/edus，从从表形如：/1/edus/11/grades）
     * @param {?} fpath
     * @return {?}
     */
    createByPath(fpath) {
        /** @type {?} */
        const pathUrl = BefDataPathUtil.convertPathToUrl(fpath);
        /** @type {?} */
        const url = `${this.baseUrl}${pathUrl}`;
        return this.request(HttpMethods.POST, url);
    }
    /**
     * 提交变更
     * @param {?} changeDetail
     * @return {?}
     */
    update(changeDetail) {
        /** @type {?} */
        const body = {
            changeDetail
        };
        /** @type {?} */
        const requestConfig = {
            body: body
        };
        return this.request(HttpMethods.PATCH, this.baseUrl, requestConfig);
    }
    /**
     * 执行保存
     * @return {?}
     */
    save() {
        return this.request(HttpMethods.PUT, this.baseUrl);
    }
    /**
     * 删除
     * @param {?} id
     * @return {?}
     */
    delete(id) {
        /** @type {?} */
        const url = `${this.baseUrl}/${id}`;
        return this.request(HttpMethods.DELETE, url);
    }
    /**
     * 删除（扩展）
     * @param {?} id
     * @return {?}
     */
    extendDelete(id) {
        /** @type {?} */
        const url = `${this.baseUrl}/extension/delete/${id}`;
        return this.request(HttpMethods.PUT, url);
    }
    /**
     * 删除并保存
     * @param {?} id
     * @return {?}
     */
    deleteAndSave(id) {
        /** @type {?} */
        const url = `${this.baseUrl}/service/delete/${id}`;
        /** @type {?} */
        const requestConfig = {};
        return this.request(HttpMethods.PUT, url, requestConfig);
    }
    /**
     * 删除后代
     * @param {?} fpath
     * @param {?} id
     * @return {?}
     */
    deletByPath(fpath, id) {
        /** @type {?} */
        const pathUrl = BefDataPathUtil.convertPathToUrl(fpath);
        /** @type {?} */
        const url = `${this.baseUrl}${pathUrl}/${id}`;
        return this.request(HttpMethods.DELETE, url);
    }
    /**
     * 删除后代（扩展）
     * @param {?} fpath
     * @param {?} id
     * @return {?}
     */
    extendDeletByPath(fpath, id) {
        /** @type {?} */
        const pathUrl = BefDataPathUtil.convertPathToUrl(fpath);
        /** @type {?} */
        const url = `${this.baseUrl}/extension${pathUrl}/${id}`;
        return this.request(HttpMethods.PUT, url);
    }
    /**
     * 批量删除
     * @param {?} ids 待删除的id数组
     * @return {?}
     */
    batchDelete(ids) {
        /** @type {?} */
        const params = {
            ids: ids.join(',')
        };
        /** @type {?} */
        const requestConfig = {
            params: params
        };
        return this.request(HttpMethods.DELETE, this.baseUrl, requestConfig);
    }
    /**
     * 批量删除（扩展）
     * @param {?} ids
     * @return {?}
     */
    extendBatchDelete(ids) {
        /** @type {?} */
        const url = `${this.baseUrl}/extension/batchdelete`;
        /** @type {?} */
        const params = {
            ids: ids.join(',')
        };
        /** @type {?} */
        const requestConfig = {
            params: params
        };
        return this.request(HttpMethods.PUT, url, requestConfig);
    }
    /**
     * 取消
     * @return {?}
     */
    cancel() {
        /** @type {?} */
        const url = `${this.baseUrl}/service/cancel`;
        return this.request(HttpMethods.POST, url);
    }
    /**
     * 调用httpclient取数
     * @param {?} method methods
     * @param {?} url url
     * @param {?=} requestConfigs options
     * @param {?=} ignoreHandlingChanges
     * @return {?}
     */
    request(method, url, requestConfigs, ignoreHandlingChanges = false) {
        requestConfigs = requestConfigs || {};
        // 扩展body
        if (ignoreHandlingChanges !== true) {
            if (method === HttpMethods.POST || method === HttpMethods.PUT || method === HttpMethods.PATCH) {
                /** @type {?} */
                const body = requestConfigs.body || {};
                requestConfigs.body = this.proxyExtend.extendBody(body);
            }
        }
        // 扩展headers
        /** @type {?} */
        const headers$ = this.proxyExtend.extendHeaders(requestConfigs.headers);
        // 发送请求
        return headers$.pipe(switchMap((/**
         * @param {?} headers
         * @return {?}
         */
        (headers) => {
            requestConfigs.headers = headers;
            requestConfigs.observe = 'response';
            return this.httpClient.request(method, url, requestConfigs).pipe(map((/**
             * @param {?} result
             * @return {?}
             */
            (result) => {
                if (ignoreHandlingChanges === true) {
                    return result && result.body && result.body.returnValue ? result.body.returnValue : result;
                }
                else {
                    return this.proxyExtend.onResponse(result);
                }
            })), catchError((/**
             * @param {?} error
             * @return {?}
             */
            (error) => {
                return this.proxyExtend.onError(error, false, false);
            })));
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            return this.proxyExtend.onError(error, false, false);
        })));
    }
}
if (false) {
    /**
     * API基路径
     * \@summary
     * 延迟到实现类确定，考虑使用注解指定。
     * @type {?}
     */
    BefProxy.prototype.baseUrl;
    /** @type {?} */
    BefProxy.prototype.associatedUrlMap;
    /**
     * 代理扩展
     * @type {?}
     * @protected
     */
    BefProxy.prototype.proxyExtend;
    /** @type {?} */
    BefProxy.prototype.httpClient;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/decorators.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 变量元数据名称
 * @type {?}
 */
const VARIABLE_PROP_META = '@farris/bef VARIABLE_PROP_META';
/**
 * NgVariable元数据接口
 * @record
 */
function VariablePropMetadata() { }
if (false) {
    /**
     * 变量影射
     * @type {?}
     */
    VariablePropMetadata.prototype.mapping;
}
/**
 * NgVariable装饰器工厂接口
 * @record
 */
function VariablePropDecorator() { }
const ɵ0 = /**
 * @param {?} obj
 * @return {?}
 */
(obj) => obj;
/**
 * NgVariable装饰工厂的工厂
 * @type {?}
 */
const VariablePropMeta = makePropDecorator(VARIABLE_PROP_META, (ɵ0));

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_entity_manager.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 实体管理类：管理Entity和
 * \@todo
 * 1、实体管理应该放在Devkit中；
 * 2、目前只是简单代码拆分，待进一步明确职责。
 * @template T
 */
class BefEntityManager extends EntityManager {
    /**
     * 构造函数
     * @param {?} entityCollection
     */
    constructor(entityCollection) {
        super(entityCollection);
        this.changeBuilder = new BefChangeBuilder(this.entityCollection.entityType, this.entityCollection);
        this.changeHandler = new BefChangeHandler();
    }
    /**
     * 获取数据变更
     * @return {?}
     */
    buildAllEntityChangeDetails() {
        /** @type {?} */
        const changeDetails = [];
        /** @type {?} */
        const entities = this.entityCollection.getAllEntities();
        entities.forEach((/**
         * @param {?} entity
         * @return {?}
         */
        (entity) => {
            if (entity.changes.length === 0) {
                return;
            }
            /** @type {?} */
            const changeDetail = this.changeBuilder.build(entity.changes);
            // 防止空id的变更被提交
            // @todo：此处判断应该放到更底层，临时修复Bug。
            if (changeDetail && changeDetail.ChangeInfo && changeDetail.ChangeInfo.DataId) {
                changeDetails.push(changeDetail);
            }
        }));
        return changeDetails;
    }
    /**
     * 获取单个实体
     * @param {?} id
     * @return {?}
     */
    buildEntityChangeDetailById(id) {
        /** @type {?} */
        const entity = this.entityCollection.getEntityById(id);
        if (entity.changes.length === 0) {
            return null;
        }
        /** @type {?} */
        const changeDetail = this.changeBuilder.build(entity.changes);
        return changeDetail;
    }
    /**
     * 应用数据变更
     * @param {?} changeDetails
     * @return {?}
     */
    handleDataChangeDetails(changeDetails) {
        this.changeHandler.handle(this.entityType, this.entityCollection, changeDetails);
    }
    // #endregion
    /**
     * 清空所有实体
     * @return {?}
     */
    reset() {
        this.entityCollection.clear();
    }
}
if (false) {
    /**
     * ChangeDetail构造器
     * @type {?}
     */
    BefEntityManager.prototype.changeBuilder;
    /**
     * ChangeDetail处理器
     * @type {?}
     */
    BefEntityManager.prototype.changeHandler;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_variable_manager.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Be变量管理器
 */
class BefVariableManager {
    /**
     * 构造函数
     * @param {?} ngVariables
     */
    constructor(ngVariables) {
        this.ngVariables = ngVariables;
        this.ngVariableMap = new Map();
        this.lastSyncValuesMap = new Map();
        this.latestVariableValues = new Map();
        // 重新组织变量元数据
        Object.keys(this.ngVariables).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        (propName) => {
            this.ngVariableMap.set(propName, ngVariables[propName]);
        }));
    }
    /**
     * 设置变量值
     * @param {?} name
     * @param {?} value
     * @return {?}
     */
    setValue(name, value) {
        if (this.ngVariableMap.has(name) === false) {
            throw new Error(`不存在名为${name}的变量`);
        }
        this.latestVariableValues.set(name, value);
    }
    /**
     * 获取变量值
     * @param {?} name
     * @return {?}
     */
    getValue(name) {
        if (this.ngVariableMap.has(name) === false) {
            throw new Error(`不存在名为${name}的变量`);
        }
        this.latestVariableValues.get(name);
    }
    /**
     * 处理服务器端返回的变量变更
     * @param {?} changeDetail
     * @return {?}
     */
    handleChangeDetail(changeDetail) {
    }
    /**
     * Build ChangeDetail instance for all variables.
     * @return {?}
     */
    buildChangeDetail() {
        /** @type {?} */
        const changeDetail = BefChangeUtil.createEmpty(ChangeDetailType.Modify);
        this.ngVariableMap.forEach((/**
         * @param {?} ngVariable
         * @param {?} varName
         * @return {?}
         */
        (ngVariable, varName) => {
            /** @type {?} */
            const newValue = this.latestVariableValues.get(varName);
            /** @type {?} */
            const oldValue = this.lastSyncValuesMap.get(varName);
            if (this.isValueEqual(newValue, oldValue) === false) {
                this.lastSyncValuesMap.set(varName, newValue);
                this.appendToChangeInfo(changeDetail, varName, newValue);
            }
        }));
        if (Object.keys(changeDetail.ChangeInfo).length === 0) {
            return null;
        }
        return changeDetail;
    }
    /**
     * Clear variable values cached in the innerValueMap property.
     * @return {?}
     */
    reset() {
        this.lastSyncValuesMap.clear();
        this.latestVariableValues.clear();
    }
    /**
     * Append changed variable to ChangeDetail instance.
     * @private
     * @param {?} changeDetail
     * @param {?} varName
     * @param {?} varValue
     * @return {?}
     */
    appendToChangeInfo(changeDetail, varName, varValue) {
        if (this.isUdtVariable(varValue) === true) {
            /** @type {?} */
            const udtVarChangeDetail = BefChangeUtil.createEmpty(ChangeDetailType.Modify);
            udtVarChangeDetail.ChangeInfo = varValue;
            changeDetail.ChangeInfo[varName] = udtVarChangeDetail;
        }
        else {
            changeDetail.ChangeInfo[varName] = varValue;
        }
    }
    /**
     * 值比较
     * \@todo 临时采用这种方式
     * @private
     * @param {?} srcValue
     * @param {?} dstValue
     * @return {?}
     */
    isValueEqual(srcValue, dstValue) {
        return JSON.stringify(srcValue) === JSON.stringify(dstValue);
    }
    /**
     * Check if the object is a plain object
     * @private
     * @param {?} obj
     * @return {?}
     */
    isUdtVariable(obj) {
        return obj && obj.constructor &&
            obj.toString() === '[object Object]' &&
            obj.constructor.prototype.hasOwnProperty('isPrototypeOf');
    }
}
if (false) {
    /**
     * 变量元数据
     * @type {?}
     * @private
     */
    BefVariableManager.prototype.ngVariableMap;
    /**
     * 最近向服务器端同步的变量值
     * @type {?}
     * @private
     */
    BefVariableManager.prototype.lastSyncValuesMap;
    /**
     * 当前最新的变量值
     * @type {?}
     * @private
     */
    BefVariableManager.prototype.latestVariableValues;
    /**
     * @type {?}
     * @private
     */
    BefVariableManager.prototype.ngVariables;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_data_service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Bef数据访问服务
 * @template T
 */
class BefDataService {
    /**
     * 构造函数
     * @param {?} repository
     */
    constructor(repository) {
        this.repository = repository;
    }
    /**
     * Api代理类
     * @private
     * @return {?}
     */
    get apiProxy() {
        return this.repository.apiProxy;
    }
    /**
     * 获取实体集合
     * @param {?} filter
     * @param {?} sort
     * @param {?} pageSize
     * @param {?} pageIndex
     * @return {?}
     */
    getList(filter, sort, pageSize, pageIndex) {
        /** @type {?} */
        const entityFilter = this.buildEntityFilter(filter, sort, pageSize, pageIndex);
        /** @type {?} */
        const query$ = this.apiProxy.extendQuery(entityFilter);
        /** @type {?} */
        const result$ = query$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        (returnValue) => {
            this.repository.setPaginationInfo(((/** @type {?} */ (returnValue))).pagination);
            /** @type {?} */
            const listData = ((/** @type {?} */ (returnValue))).result;
            /** @type {?} */
            const entities = this.repository.buildEntities(listData);
            if (this.determineIsAppend(pageIndex)) {
                this.repository.entityCollection.addEntities(entities);
            }
            else {
                this.repository.entityCollection.loadEntities(entities);
            }
            return entities;
        })));
        return result$;
    }
    /**
     * 根据当前页判断是否追加
     * @param {?} pageIndex
     * @return {?}
     */
    determineIsAppend(pageIndex) {
        return pageIndex > 1;
    }
    /**
     * 获取单个实体
     * @param {?} id
     * @return {?}
     */
    getById(id) {
        /** @type {?} */
        const retrieve$ = this.apiProxy.extendRetrieve(id);
        /** @type {?} */
        const result$ = retrieve$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        (returnValue) => {
            /** @type {?} */
            const entityData = returnValue;
            /** @type {?} */
            const entity = this.repository.buildEntity(entityData);
            this.repository.entityCollection.loadEntities([entity]);
            return entity;
        })));
        return result$;
    }
    /**
     * 更新并对实体加锁
     * @param {?} id
     * @return {?}
     */
    editById(id) {
        /** @type {?} */
        const entity = this.repository.entityCollection.getEntityById(id);
        if (!entity) {
            return of(null);
        }
        /** @type {?} */
        const retrieve$ = this.apiProxy.edit(id);
        /** @type {?} */
        const result$ = retrieve$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        (returnValue) => {
            /** @type {?} */
            const entityData = returnValue.data;
            /** @type {?} */
            const entityToUpdate = this.repository.entityCollection.getEntityById(id);
            if (entityToUpdate && entityData) {
                this.reloadEntityData(entityToUpdate, entityData);
            }
            return entityToUpdate;
        })));
        return result$;
    }
    /**
     * 根据id更新实体
     * @param {?} id 实体id
     * @return {?}
     */
    updateById(id) {
        /** @type {?} */
        const entity = this.repository.entityCollection.getEntityById(id);
        if (!entity) {
            return EMPTY;
        }
        /** @type {?} */
        const retrieve$ = this.apiProxy.extendRetrieve(id);
        /** @type {?} */
        const result$ = retrieve$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        (returnValue) => {
            /** @type {?} */
            const entityData = returnValue;
            /** @type {?} */
            const entityToUpdate = this.repository.entityCollection.getEntityById(id);
            this.reloadEntityData(entityToUpdate, entityData);
            return entityToUpdate;
        })));
        return result$;
    }
    /**
     * 重新加载实体数据
     * @private
     * @param {?} entity
     * @param {?} entityData
     * @return {?}
     */
    reloadEntityData(entity, entityData) {
        if (!entity) {
            return;
        }
        entity.load(entityData);
        entity.changes.splice(0, entity.changes.length);
    }
    /**
     * 创建新实体，并加载
     * @param {?=} defaultValue
     * @return {?}
     */
    create(defaultValue) {
        /** @type {?} */
        const result$ = this.apiProxy.create(defaultValue);
        return result$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        (returnValue) => {
            /** @type {?} */
            const newEntityData = returnValue;
            /** @type {?} */
            const newEntity = this.repository.buildEntity(newEntityData);
            this.repository.entityCollection.loadEntities([newEntity]);
            this.repository.dataChangeHistory.addChange({ dataId: newEntity.primaryValue, changeType: DataChangeType.Add });
            return newEntity;
        })));
    }
    /**
     * 追加实体
     * @param {?=} defaultValue
     * @return {?}
     */
    append(defaultValue) {
        /** @type {?} */
        const append$ = this.apiProxy.create(defaultValue);
        return append$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        (returnValue) => {
            /** @type {?} */
            const newEntityData = returnValue;
            /** @type {?} */
            const newEntity = this.repository.buildEntity(newEntityData);
            this.repository.entityCollection.addEntity(newEntity);
            return newEntity;
        })));
    }
    /**
     * 创建子实体，并追加
     * @param {?} fpath 父路径，格式形如：/1/edus（从表）或/1/edus/2/grades（从从表）
     * @return {?}
     */
    appendByPath(fpath) {
        /** @type {?} */
        const append$ = this.apiProxy.createByPath(fpath);
        return append$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        (returnValue) => {
            /** @type {?} */
            const newEntityData = returnValue;
            /** @type {?} */
            const newEntity = this.repository.entityManager.appendEntityByPath(fpath, newEntityData, newEntityData);
            return newEntity;
        })));
    }
    /**
     * 根据id删除实体
     * @param {?} id 内码
     * @param {?=} ifSave
     * @return {?}
     */
    removeById(id, ifSave) {
        ifSave = (ifSave === undefined) ? true : ifSave;
        /** @type {?} */
        const delete$ = this.apiProxy.extendDelete(id);
        // 从本地实体集合中移除
        return delete$.pipe(switchMap((/**
         * @return {?}
         */
        () => {
            // 执行保存
            if (ifSave === true) {
                return this.applyChangesById(id).pipe(tap((/**
                 * @param {?} result
                 * @return {?}
                 */
                result => {
                    if (result) {
                        this.repository.entityCollection.removeEntityById(id);
                    }
                })));
            }
            else {
                // 从实体集合中删除
                this.repository.entityCollection.removeEntityById(id);
                this.repository.dataChangeHistory.addChange({ dataId: id, changeType: DataChangeType.Delete });
                return of(true);
            }
        })));
    }
    /**
     * 删除并保存
     * @param {?} id 要删除的实体id
     * @return {?}
     */
    removeAndSaveById(id) {
        /** @type {?} */
        const delete$ = this.apiProxy.deleteAndSave(id);
        /** @type {?} */
        const result$ = delete$.pipe(switchMap((/**
         * @return {?}
         */
        () => {
            this.repository.entityCollection.removeEntityById(id);
            return of(true);
        })));
        return result$;
    }
    /**
     * 批量删除
     * @param {?} ids
     * @return {?}
     */
    removeByIds(ids) {
        // 服务器端删除
        /** @type {?} */
        const delete$ = this.apiProxy.extendBatchDelete(ids);
        // 从本地实体集合中移除
        return delete$.pipe(switchMap((/**
         * @return {?}
         */
        () => {
            // 执行保存
            return this.applyChangesByIdArray(ids).pipe(tap((/**
             * @param {?} result
             * @return {?}
             */
            result => {
                if (result) {
                    // 从实体集合中删除
                    this.repository.entityCollection.removeEntities((/**
                     * @param {?} value
                     * @return {?}
                     */
                    (value) => {
                        return this.checkEntityValueExists(value, ids);
                    }));
                }
            })));
        })));
    }
    /**
     * 检测实体的主键值是否存在于数组中，如果存在返回true
     * @private
     * @param {?} entity
     * @param {?} primaryValueArray
     * @return {?}
     */
    checkEntityValueExists(entity, primaryValueArray) {
        /** @type {?} */
        let isExistInTargetArray = false;
        for (let i = 0; i < primaryValueArray.length; i++) {
            if (entity.primaryValue === primaryValueArray[i]) {
                isExistInTargetArray = true;
                break;
            }
        }
        return isExistInTargetArray;
    }
    /**
     * 删除子级
     * @param {?} fpath
     * @param {?} id   内码
     * @return {?}
     */
    removeByPath(fpath, id) {
        /** @type {?} */
        const delete$ = this.apiProxy.extendDeletByPath(fpath, id);
        // 从本地实体集合中移除
        return delete$.pipe(map((/**
         * @return {?}
         */
        () => {
            this.repository.entityManager.removeEntityByPath(fpath, id);
            this.repository.dataChangeHistory.addChange({ fpath, dataId: id, changeType: DataChangeType.Delete });
            return true;
        })));
    }
    /**
     * 将id对应的实体的变更提交的服务器端
     * @param {?} id
     * @return {?}
     */
    updateChangesById(id) {
        /** @type {?} */
        const entity = this.repository.entityCollection.getEntityById(id);
        // 变更集为空
        if (!entity.changes || entity.changes.length === 0) {
            return of(true);
        }
        /** @type {?} */
        const changeDetail = this.repository.entityManager.buildEntityChangeDetailById(id);
        /** @type {?} */
        const update$ = this.apiProxy.update(changeDetail);
        /** @type {?} */
        const result$ = update$.pipe(tap((/**
         * @return {?}
         */
        () => {
            this.repository.entityManager.clearEntityChangesById(id);
        })), map((/**
         * @return {?}
         */
        () => {
            return true;
        })));
        return result$;
    }
    /**
     * 提交所有变更
     * @return {?}
     */
    updateAllChanges() {
        // 遍历实体，提交变更
        /** @type {?} */
        const updateResults = [];
        /** @type {?} */
        const entities = this.repository.entityCollection.toArray();
        if (entities.length === 0) {
            return of(true);
        }
        entities.forEach((/**
         * @param {?} entity
         * @return {?}
         */
        (entity) => {
            /** @type {?} */
            const updateResult$ = this.updateChangesById(entity.primaryValue);
            updateResults.push(updateResult$);
        }));
        // 串联流
        /** @type {?} */
        const result$ = zip(...updateResults).pipe(map((/**
         * @return {?}
         */
        () => {
            return true;
        })));
        return result$;
    }
    /**
     * 应用变更
     * @return {?}
     */
    applyChanges() {
        /** @type {?} */
        const save$ = this.apiProxy.save();
        /** @type {?} */
        const result$ = save$.pipe(tap((/**
         * @return {?}
         */
        () => {
            this.repository.entityManager.clearAllEntityChanges();
            this.repository.dataChangeHistory.clear();
        })), map((/**
         * @return {?}
         */
        () => {
            return true;
        })));
        return result$;
    }
    /**
     * 应用idArray对应数据的变更
     * @param {?} ids
     * @return {?}
     */
    applyChangesByIdArray(ids) {
        /** @type {?} */
        const save$ = this.apiProxy.save();
        /** @type {?} */
        const result$ = save$.pipe(tap((/**
         * @return {?}
         */
        () => {
            this.repository.entityManager.clearEntityChangesByIds(ids);
            this.repository.dataChangeHistory.clearByIds(ids);
        })), map((/**
         * @return {?}
         */
        () => {
            return true;
        })));
        return result$;
    }
    /**
     * 应用id对应数据的变更
     * @param {?} id
     * @return {?}
     */
    applyChangesById(id) {
        /** @type {?} */
        const save$ = this.apiProxy.save();
        /** @type {?} */
        const result$ = save$.pipe(tap((/**
         * @return {?}
         */
        () => {
            this.repository.entityManager.clearEntityChangesById(id);
            this.repository.dataChangeHistory.clearByIds([id]);
        })), map((/**
         * @return {?}
         */
        () => {
            return true;
        })));
        return result$;
    }
    /**
     * 取消变更
     * @return {?}
     */
    cancelChanges() {
        /** @type {?} */
        const cancel$ = this.apiProxy.cancel();
        /** @type {?} */
        const result$ = cancel$.pipe(tap((/**
         * @return {?}
         */
        () => {
            this.repository.entityManager.clearAllEntityChanges();
            this.repository.dataChangeHistory.clear();
        })), map((/**
         * @return {?}
         */
        () => {
            return true;
        })));
        return result$;
    }
    /**
     * 构造EntityFilter
     * @private
     * @param {?} filter
     * @param {?} sort
     * @param {?} pageSize
     * @param {?} pageIndex
     * @return {?}
     */
    buildEntityFilter(filter, sort, pageSize, pageIndex) {
        /** @type {?} */
        const entityFilter = {
            FilterConditions: filter,
            SortConditions: sort,
            IsUsePagination: pageSize === 0 ? false : true,
            Pagination: {
                PageIndex: pageIndex,
                PageSize: pageSize,
                PageCount: 0,
                TotalCount: 0
            }
        };
        return entityFilter;
    }
}
if (false) {
    /**
     * 实体仓库
     * @type {?}
     * @private
     */
    BefDataService.prototype.repository;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/framework_session_service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 框架Session服务
 */
class FrameworkSessionService {
    constructor() {
    }
    /**
     * 获取用户SessionId
     * @return {?}
     */
    getUserSessionId() {
        /** @type {?} */
        const userSessionId = window.localStorage.getItem('sessionId');
        return userSessionId;
    }
    /**
     * 获取框架菜单id
     * @return {?}
     */
    getFuncSessionId() {
        /** @type {?} */
        const frmMobileService = window['frmMobileService'];
        if (frmMobileService) {
            return frmMobileService.rtf.commonVariable.formToken();
        }
        return null;
    }
    /**
     * 获取当前会话id
     * @return {?}
     */
    getCurrentSessionId() {
        /** @type {?} */
        let funcSessionId = window.sessionStorage.getItem('FuncSessionId');
        /** @type {?} */
        let latestFuncSessionId = this.getFuncSessionId();
        if (latestFuncSessionId && latestFuncSessionId !== funcSessionId) {
            // 设置为空，触发后续逻辑更新最新FuncSessionId
            funcSessionId = null;
        }
        if (!funcSessionId) {
            funcSessionId = this.getFuncSessionId();
            if (funcSessionId) {
                window.sessionStorage.setItem('FuncSessionId', funcSessionId);
            }
        }
        return funcSessionId;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_proxy_extend.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 代理钩子
 * @record
 */
function IProxyExtend() { }
if (false) {
    /**
     * @param {?} body
     * @return {?}
     */
    IProxyExtend.prototype.extendBody = function (body) { };
    /**
     * @param {?} headers
     * @return {?}
     */
    IProxyExtend.prototype.extendHeaders = function (headers) { };
    /**
     * @param {?} response
     * @param {?=} ignoreChanges
     * @return {?}
     */
    IProxyExtend.prototype.onResponse = function (response, ignoreChanges) { };
    /**
     * @param {?} error
     * @param {?} selfHandError
     * @param {?} ignoreError
     * @return {?}
     */
    IProxyExtend.prototype.onError = function (error, selfHandError, ignoreError) { };
}
/**
 * Bef代理扩展类
 */
class BefProxyExtend {
    /**
     * 构造函数
     * @param {?} repository
     */
    constructor(repository) {
        this.repository = repository;
    }
    /**
     * 注入器
     * @private
     * @return {?}
     */
    get injector() {
        return this.repository.getInjector();
    }
    /**
     * 返回结果处理
     * @param {?} response
     * @return {?}
     */
    onResponse(response) {
        this.handleResponseHeaders(response.headers);
        return this.handleResponseBody(response.body);
    }
    /**
     * 处理服务器端返回的headers数据
     * @private
     * @param {?} headers
     * @return {?}
     */
    handleResponseHeaders(headers) {
        this.repository.sessionService.handleResponseHeaders(headers);
    }
    /**
     * 处理服务器端返回的的body数据
     * @private
     * @param {?} responseInfo
     * @return {?}
     */
    handleResponseBody(responseInfo) {
        if (responseInfo && responseInfo.innerDataChange) {
            this.repository.entityManager.handleDataChangeDetails(responseInfo.innerDataChange);
        }
        this.repository.entityManager.clearAllEntityChanges();
        if (responseInfo && responseInfo.hasOwnProperty('returnValue')) {
            return responseInfo.returnValue;
        }
        else {
            return responseInfo;
        }
    }
    /**
     * 错误处理
     * @param {?} error
     * @return {?}
     */
    onError(error) {
        // 获取所有load组件实例，消除
        /** @type {?} */
        const loadingServices = window['DEVKIT_LOADING_SERVICE'];
        if (loadingServices && loadingServices instanceof Array && loadingServices.length > 0) {
            for (const loadingService of loadingServices) {
                if (typeof (loadingService.hide) === 'function') {
                    loadingService.hide();
                }
            }
        }
        // 捕获到异常，处理异常信息
        /** @type {?} */
        const exceptionService = this.injector.get(BE_ERROR_HANDLER__TOKEN, null);
        if (!!exceptionService) {
            exceptionService.show(error);
            return EMPTY;
        }
        else {
            return EMPTY;
        }
    }
    /**
     * 扩展Headers
     * @param {?} headers
     * @return {?}
     */
    extendHeaders(headers) {
        /** @type {?} */
        const $getSessionId = this.repository.sessionService.getBeSessionId();
        return $getSessionId.pipe(switchMap((/**
         * @return {?}
         */
        () => {
            headers = this.repository.sessionService.extendRequestHeaders(headers);
            return of(headers);
        })));
    }
    /**
     * 扩展Body
     * @param {?} body
     * @return {?}
     */
    extendBody(body) {
        return RequestInfoUtil.appendRequestInfoToBody(body, this.repository);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    BefProxyExtend.prototype.repository;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_repository_initializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * BefRepository初始化器
 * \@summary
 * 为了抽离这部分路径，强制对外暴露了Repository的相关属性，待优化。
 */
class BefRepositoryInitializer {
    /**
     * 注入器
     * @private
     * @return {?}
     */
    get injector() {
        return this.repository.getInjector();
    }
    /**
     * 构造函数
     * @param {?} repository
     */
    constructor(repository) {
        this.repository = repository;
    }
    /**
     * 初始化
     * @return {?}
     */
    init() {
        this.initAppContext();
        this.initEntityManager();
        this.initVariableManager();
        this.initApiProxy();
        this.initSessionService();
        this.initDataService();
    }
    /**
     * 初始化应用上下文
     * @private
     * @return {?}
     */
    initAppContext() {
        this.repository.appContext = this.injector.get(AppContext);
    }
    /**
     * 初始化EntityManager
     * @private
     * @return {?}
     */
    initEntityManager() {
        this.repository.entityManager = new BefEntityManager(this.repository.entityCollection);
    }
    /**
     * 初始化VariableManager
     * @private
     * @return {?}
     */
    initVariableManager() {
        /** @type {?} */
        const ngVariables = MetadataUtil.getPropsMetadatasByName(this.repository.constructor, VARIABLE_PROP_META);
        this.repository.variableManager = new BefVariableManager(ngVariables);
    }
    /**
     * 初始化Proxy
     * @private
     * @return {?}
     */
    initApiProxy() {
        this.repository.apiProxy = this.injector.get(this.repository.apiProxyType);
        /** @type {?} */
        const apiProxyExtend = new BefProxyExtend(this.repository);
        this.repository.apiProxy.setProxyExtend(apiProxyExtend);
    }
    /**
     * 初始化SessionService
     * @private
     * @return {?}
     */
    initSessionService() {
        /** @type {?} */
        let handlingStrategyName = this.injector.get(BE_SESSION_HANDLING_STRATEGY_TOKEN, null, InjectFlags.Optional);
        handlingStrategyName = handlingStrategyName || 'SeparatedSession';
        /** @type {?} */
        const frmSessionService = this.injector.get(FrameworkSessionService);
        /** @type {?} */
        const httpClient = this.injector.get(HttpClient);
        /** @type {?} */
        const baseUrl = `${this.repository.apiProxy.baseUrl}`;
        /** @type {?} */
        const sessionHandlingStrategyFactory = new BefSessionHandlingStrategyFactory();
        /** @type {?} */
        const sessionHandlingStrategy = sessionHandlingStrategyFactory.create(handlingStrategyName, frmSessionService, baseUrl, httpClient);
        this.repository.sessionService = new BefSessionService(sessionHandlingStrategy);
    }
    /**
     * 初始化DataService
     * @private
     * @return {?}
     */
    initDataService() {
        this.repository.dataService = new BefDataService(this.repository);
    }
}
if (false) {
    /**
     * BefRepository实例
     * @type {?}
     * @private
     */
    BefRepositoryInitializer.prototype.repository;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_repository.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Bef实体仓库实现类
 * @abstract
 * @template T
 */
class BefRepository extends Repository {
    /**
     * 构造函数
     * @param {?} injector
     */
    constructor(injector) {
        super();
        this.injector = injector;
    }
    /**
     * 初始化
     * \@summary
     * 实现类的构造函数中调用才能正确获取apiProxy
     * @protected
     * @return {?}
     */
    init() {
        super.init();
        /** @type {?} */
        const initializer = new BefRepositoryInitializer(this);
        initializer.init();
    }
    /**
     * 获取注入器
     * @return {?}
     */
    getInjector() {
        return this.injector;
    }
    /**
     * 获取实体数组
     * @param {?} filter
     * @param {?} sort
     * @param {?} pageSize
     * @param {?} pageIndex
     * @return {?}
     */
    getEntities(filter, sort, pageSize, pageIndex) {
        return this.dataService.getList(filter, sort, pageSize, pageIndex);
    }
    /**
     * 获取主键为id的实体
     * @param {?} id 实体id
     * @return {?}
     */
    getEntityById(id) {
        return this.dataService.getById(id);
    }
    /**
     * 更新实体数据
     * @param {?} id
     * @return {?}
     */
    updateEntityById(id) {
        return this.dataService.updateById(id);
    }
    /**
     * 更新实体数据并加锁
     * @param {?} id
     * @return {?}
     */
    editEntityById(id) {
        return this.dataService.editById(id);
    }
    /**
     * 创建实体
     * @param {?=} defaultValue
     * @return {?}
     */
    createEntity(defaultValue) {
        return this.dataService.create(defaultValue);
    }
    /**
     * 追加实体
     * @param {?=} defaultValue
     * @return {?}
     */
    appendEntity(defaultValue) {
        return this.dataService.append(defaultValue);
    }
    /**
     * 创建子实体，并追加
     * @param {?} fpath 父路径，格式形如：/1/edus（从表）或/1/edus/2/grades（从从表）
     * @return {?}
     */
    appendEntityByPath(fpath) {
        return this.dataService.appendByPath(fpath);
    }
    /**
     * 根据id删除实体
     * @param {?} id
     * @param {?=} ifSave
     * @return {?}
     */
    removeEntityById(id, ifSave) {
        return this.dataService.removeById(id, ifSave);
    }
    /**
     * 根据ids批量删除实体
     * @param {?} ids
     * @return {?}
     */
    removeEntitiesByIds(ids) {
        return this.dataService.removeByIds(ids);
    }
    /**
     * 根据id删除实体并执行保存
     * @param {?} id
     * @return {?}
     */
    removeEntityAndSaveById(id) {
        return this.dataService.removeAndSaveById(id);
    }
    /**
     * 删除子级
     * @param {?} fpath
     * @param {?} id
     * @return {?}
     */
    removeEntityByPath(fpath, id) {
        return this.dataService.removeByPath(fpath, id);
    }
    /**
     * 保存主键为id的实体
     * @param {?} id
     * @return {?}
     */
    saveEntityById(id) {
        return this.dataService.applyChangesById(id);
    }
    /**
     * 批量保存
     * @return {?}
     */
    saveEntities() {
        return this.dataService.applyChanges();
    }
    /**
     * 取消实体变更
     * @return {?}
     */
    cancelEntityChanges() {
        return this.dataService.cancelChanges();
    }
    /**
     * 清空Repositoy内的数据和变量
     * @return {?}
     */
    reset() {
        this.entityManager.reset();
        this.variableManager.reset();
        this.sessionService.clearBeSessionId();
    }
}
if (false) {
    /**
     * 名称
     * @type {?}
     */
    BefRepository.prototype.name;
    /**
     * Bef代理
     * @type {?}
     */
    BefRepository.prototype.apiProxyType;
    /**
     * 应用上下文
     * @type {?}
     */
    BefRepository.prototype.appContext;
    /**
     * 实体管理器
     * @type {?}
     */
    BefRepository.prototype.entityManager;
    /**
     * 变量管理器
     * @type {?}
     */
    BefRepository.prototype.variableManager;
    /**
     * API代理
     * @type {?}
     */
    BefRepository.prototype.apiProxy;
    /**
     * Bef会话管理
     * @type {?}
     */
    BefRepository.prototype.sessionService;
    /**
     * Bef数据操作
     * @type {?}
     */
    BefRepository.prototype.dataService;
    /**
     * @type {?}
     * @protected
     */
    BefRepository.prototype.injector;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef-lookup-data.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Bef帮助取数服务
 */
class BefLookupDataService {
    /**
     * 构造函数
     * @param {?} viewModelContext
     */
    constructor(viewModelContext) {
        this.viewModelContext = viewModelContext;
        this.befRepository = (/** @type {?} */ (this.viewModelContext.repository));
    }
    /**
     * 帮助取数
     * @param {?} helpMetadataId
     * @param {?=} queryParam
     * @return {?}
     */
    getData(helpMetadataId, queryParam) {
        /** @type {?} */
        const tableName = helpMetadataId.split('.')[0];
        /** @type {?} */
        const labelId = helpMetadataId.split('.')[1];
        queryParam = queryParam || {};
        return this.extendGetHelpData(labelId, tableName, queryParam);
    }
    /**
     * 扩展的帮助取数
     * @private
     * @param {?} labelId
     * @param {?} tableName
     * @param {?} queryParam
     * @return {?}
     */
    extendGetHelpData(labelId, tableName, queryParam) {
        /** @type {?} */
        const url = `${this.befRepository.apiProxy.baseUrl}/extension/elementhelps`;
        /** @type {?} */
        const requestInfo = RequestInfoUtil.buildRequestInfo(this.befRepository);
        /** @type {?} */
        const body = {
            labelId: labelId,
            nodeCode: tableName,
            queryParam: queryParam,
            requestInfo: requestInfo
        };
        /** @type {?} */
        const requestConfig = {
            body: body
        };
        /** @type {?} */
        const result$ = this.befRepository.apiProxy.request(HttpMethods.PUT, url, requestConfig);
        return result$.pipe(map((/**
         * @param {?} result
         * @return {?}
         */
        (result) => {
            return result;
        })));
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    BefLookupDataService.prototype.viewModelContext;
    /**
     * @type {?}
     * @private
     */
    BefLookupDataService.prototype.befRepository;
    /** @type {?} */
    BefLookupDataService.prototype.context;
}

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: farris-mobile-bef.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { BE_ERROR_HANDLER__TOKEN, BE_SERVER_URI_TOKEN, BE_SESSION_HANDLING_STRATEGY_TOKEN, BefChangeBuilder, BefChangeHandler, BefChangeUtil, BefDataPathUtil, BefEnvUtil, BefHttpUtil, BefLookupDataService, BefProxy, BefProxyUtil, BefRepository, BefRepositoryUtil, BefSeparatedSessionHandlingStrategy, BefSessionHandlingStrategy, BefSessionHandlingStrategyFactory, BefSessionService, BefUnifiedSessionHandlingStrategy, ChangeDetailType, EntityUtil, FrameworkSessionService, RequestInfoUtil, ResponseInfoUtil, SessionStorageBeSessionStorageStrategy, VARIABLE_PROP_META, VariablePropMeta };
