// @farris/devkit/esm2015/lib/form/decorators.js
import { makePropDecorator, makeDecorator } from '../metadata/index';
/**
 * ----------------------------------------
 * NgValidateForm
 * ----------------------------------------
 */
/**
 * 子表单装饰器名称
 */
export const NG_VALIDATE_FORM = 'NgValidateForm';
/**
 * 表单装饰器工厂
 */
export function NgValidateForm(options) {
    const decoratorFactory = makeDecorator(NG_VALIDATE_FORM, (obj) => obj);
    return decoratorFactory(options);
}
/**
 * ----------------------------------------
 * NgChildForm
 * ----------------------------------------
 */
/**
 * 子表单装饰器名称
 */
export const NG_CHILD_FORM = 'NgChildForm';
/**
 * 子表单装饰器工厂
 */
export const NgChildForm = makePropDecorator(NG_CHILD_FORM, (obj) => obj);
/**
 * ----------------------------------------
 * NgChildFormArray
 * ----------------------------------------
 */
/**
 * 子表单数组装饰器名称
 * @type {string}
 */
export const NG_CHILD_FORM_ARRAY = 'NgChildFormArray';
/**
 * 子表单数组装饰器工厂
 */
export const NgChildFormArray = makePropDecorator(NG_CHILD_FORM_ARRAY, (obj) => obj);
/**
 * ----------------------------------------
 * NgFormControl
 * ----------------------------------------
 */
/**
 * 表单控件装饰器名称
 * @type {string}
 */
export const NG_FORM_CONTROL = 'NgFormControl';
/**
 * 表单控件装饰器工厂
 */
export const NgFormControl = makePropDecorator(NG_FORM_CONTROL, (obj) => obj);
