// @edp-pmf/bpmn-model/dist/cjs/lib/model/mex/ConstantExpression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstantExpression = void 0;
var tslib_1 = require("tslib");
var BpmnModelHelper_1 = require("../BpmnModelHelper");
var MexExpression_1 = require("./MexExpression");
var ConstantExpression = /** @class */ (function (_super) {
    tslib_1.__extends(ConstantExpression, _super);
    function ConstantExpression(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.Mex.ConstantExpression";
        return _this;
    }
    ConstantExpression.prototype.NewInstance = function () {
        return new ConstantExpression(this.Model);
    };
    ConstantExpression.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Value = null;
    };
    ConstantExpression.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Value", this.Value);
        return obj;
    };
    ConstantExpression.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Value = obj.Value;
    };
    return ConstantExpression;
}(MexExpression_1.MexExpression));
exports.ConstantExpression = ConstantExpression;
