// @edp-pmf/bpmn-model/dist/cjs/lib/model/mex/CustomType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomType = void 0;
var tslib_1 = require("tslib");
var BpmnModelHelper_1 = require("../BpmnModelHelper");
var MexElement_1 = require("./MexElement");
var CustomType = /** @class */ (function (_super) {
    tslib_1.__extends(CustomType, _super);
    function CustomType(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.Mex.CustomType";
        return _this;
    }
    CustomType.prototype.NewInstance = function () {
        return new CustomType(this.Model);
    };
    CustomType.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.AssemblyName = null;
        this.TypeName = null;
    };
    CustomType.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "AssemblyName", this.AssemblyName);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "TypeName", this.TypeName);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "ClrTypeID", this.ClrTypeID);
        return obj;
    };
    CustomType.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.AssemblyName = obj.AssemblyName;
        this.TypeName = obj.TypeName;
    };
    CustomType.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
    };
    return CustomType;
}(MexElement_1.MexElement));
exports.CustomType = CustomType;
