// @edp-pmf/bpmn-model/dist/cjs/lib/model/mex/ExecutionListener.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExecutionListener = void 0;
var tslib_1 = require("tslib");
var BpmnModelHelper_1 = require("../BpmnModelHelper");
var MexExtensionElement_1 = require("./MexExtensionElement");
var ExecutionListener = /** @class */ (function (_super) {
    tslib_1.__extends(ExecutionListener, _super);
    function ExecutionListener(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.Mex.ExecutionListener";
        return _this;
    }
    ExecutionListener.prototype.NewInstance = function () {
        return new ExecutionListener(this.Model);
    };
    ExecutionListener.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Class = null;
        this.Event = null;
        ;
    };
    ExecutionListener.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Event", this.Event);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Class", this.Class);
        return obj;
    };
    ExecutionListener.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Event = obj.Event;
        this.Class = obj.Class;
    };
    return ExecutionListener;
}(MexExtensionElement_1.MexExtensionElement));
exports.ExecutionListener = ExecutionListener;
