// @edp-pmf/bpmn-model/dist/cjs/lib/model/mex/VariableExpression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VariableExpression = void 0;
var tslib_1 = require("tslib");
var BpmnModelHelper_1 = require("../BpmnModelHelper");
var MexExpression_1 = require("./MexExpression");
var VariableExpressionDirection_1 = require("./VariableExpressionDirection");
var VariableExpression = /** @class */ (function (_super) {
    tslib_1.__extends(VariableExpression, _super);
    function VariableExpression(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.Mex.VariableExpression";
        _this.Direction = VariableExpressionDirection_1.VariableExpressionDirection.ActivityDataInputAssociation;
        return _this;
    }
    VariableExpression.prototype.NewInstance = function () {
        return new VariableExpression(this.Model);
    };
    VariableExpression.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Direction = null;
        this.SourceRef = null;
    };
    VariableExpression.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "SourceRef", this.SourceRef);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Direction", this.Direction);
        return obj;
    };
    VariableExpression.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.SourceRef = obj.SourceRef;
        this.Direction = obj.Direction;
    };
    return VariableExpression;
}(MexExpression_1.MexExpression));
exports.VariableExpression = VariableExpression;
