// @gsp-wf/wf-biz-navigation/esm5/lib/wf-biz-navigation.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Injector, Output, EventEmitter, ViewChild, Input } from '@angular/core';
import { BillcatTreeComponent } from '@gsp-bef/billcat-tree';
import { HttpService } from '@ecp-caf/caf-common';
import { TranslatePipe } from './services/translate.pipe';
var WfBizNavigationComponent = /** @class */ (function () {
    function WfBizNavigationComponent(injector, http, translate) {
        this.injector = injector;
        this.http = http;
        this.translate = translate;
        this.selectNodeEmitter = new EventEmitter();
        this.bizObject = new EventEmitter();
        this.bizCategory = new EventEmitter();
        this.tab = 'billCategory';
        this.data = new Array();
        this.config = {
            expandLevel: 1
        };
        this.emptyData = new Array();
        this.emptyCols = [{
                field: 'code', width: 100, title: this.translate.transform('static.code'),
            }];
    }
    /**
     * @return {?}
     */
    WfBizNavigationComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.getBillcategoriesByExtend();
    };
    /**
     * @return {?}
     */
    WfBizNavigationComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        // this.billcat.treeConfig.expandLevel = 1;
    };
    /**
     * @param {?} args
     * @return {?}
     */
    WfBizNavigationComponent.prototype.tabChange = /**
     * @param {?} args
     * @return {?}
     */
    function (args) {
        this.tab = args.nextId;
        // console.log({ tabchange: args });
    };
    /**
     * @param {?} data
     * @return {?}
     */
    WfBizNavigationComponent.prototype.boTreeSelectionChanged = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data.data && data.data.isDetail === '1') {
            /** @type {?} */
            var node = {
                type: 'bo',
                nodeInfo: data
            };
            this.bizObject.emit(node);
            this.selectNodeEmitter.emit(node);
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    WfBizNavigationComponent.prototype.billCategorySelectionChanged = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data && data.detail) {
            /** @type {?} */
            var node = {
                type: 'bill',
                nodeInfo: data
            };
            this.bizCategory.emit(node);
            this.selectNodeEmitter.emit(node);
        }
    };
    /**
     * @private
     * @return {?}
     */
    WfBizNavigationComponent.prototype.getBillcategoriesByExtend = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var url = '/api/runtime/bcc/v1.0/billcategory/billcategoriesByExtend?allextends=false&customconfig=processAssignManager';
        this.http.get(url).subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            if (re) {
                _this.data = re;
            }
        }));
    };
    WfBizNavigationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-biz-navigation',
                    template: "\r\n<farris-tabs #tabs [value]=\"tab\" [fill]=\"true\" [contentFill]=\"true\" (tabChange)=\"tabChange($event)\"\r\n    class=\"f-utils-fill h-100\" style=\"background: #fff;\">\r\n    <farris-tab [title]=\"'static.kind' | translate\" [id]=\"'billCategory'\" class=\"f-utils-fill h-100\"><!--title=\"\u4E1A\u52A1\u79CD\u7C7B\"-->\r\n        <ng-template #content>\r\n            <div *ngIf=\"data.length>0; else empty\" class=\"f-utils-flex-column h-100\">\r\n                <perfect-scrollbar [config]=\"{}\">\r\n                    <billcat-tree [data]=\"data\" [config]=\"config\"\r\n                        (selectNodeEmitter)=\"billCategorySelectionChanged($event)\">\r\n                    </billcat-tree>\r\n                </perfect-scrollbar>\r\n            </div>\r\n        </ng-template>\r\n    </farris-tab>\r\n    <farris-tab [title] = \"'static.object' | translate\" [id]=\"'bizObject'\" class=\"f-utils-fill h-100\"><!--title=\"\u4E1A\u52A1\u5BF9\u8C61\"-->\r\n        <ng-template #content>\r\n            <div class = \"f-utils-flex-column h-100\" style=\"position: relative;\">\r\n                <app-bo-tree-panel (selectionChanged)=\"boTreeSelectionChanged($event)\" [showIcon]=\"true\">\r\n                </app-bo-tree-panel>\r\n            </div>\r\n        </ng-template>\r\n    </farris-tab>\r\n</farris-tabs>\r\n\r\n<ng-template #empty>\r\n    <div class=\"f-utils-fill-flex-column h-100\">\r\n        <div class=\"f-datagrid-norecords\" style=\"height: 100%;\">\r\n            <div class=\"f-datagrid-norecords-content\" style=\"width: 310px;\">{{'static.noData' | translate}}</div>\r\n        </div>\r\n    </div>\r\n</ng-template>\r\n\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    WfBizNavigationComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: HttpService },
        { type: TranslatePipe }
    ]; };
    WfBizNavigationComponent.propDecorators = {
        selectNodeEmitter: [{ type: Output }],
        bizObject: [{ type: Output }],
        bizCategory: [{ type: Output }],
        tab: [{ type: Input }],
        billcat: [{ type: ViewChild, args: [BillcatTreeComponent,] }]
    };
    return WfBizNavigationComponent;
}());
export { WfBizNavigationComponent };
if (false) {
    /** @type {?} */
    WfBizNavigationComponent.prototype.selectNodeEmitter;
    /** @type {?} */
    WfBizNavigationComponent.prototype.bizObject;
    /** @type {?} */
    WfBizNavigationComponent.prototype.bizCategory;
    /** @type {?} */
    WfBizNavigationComponent.prototype.tab;
    /** @type {?} */
    WfBizNavigationComponent.prototype.billcat;
    /** @type {?} */
    WfBizNavigationComponent.prototype.data;
    /** @type {?} */
    WfBizNavigationComponent.prototype.config;
    /** @type {?} */
    WfBizNavigationComponent.prototype.emptyData;
    /** @type {?} */
    WfBizNavigationComponent.prototype.emptyCols;
    /**
     * @type {?}
     * @private
     */
    WfBizNavigationComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    WfBizNavigationComponent.prototype.http;
    /**
     * @type {?}
     * @private
     */
    WfBizNavigationComponent.prototype.translate;
}
