// @gspwidget/widget-devkit/esm2015/lib/widget-registry/widget-event.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { curLang, FILTERS_WIDGET_FILTERS_PROP_NAME, FILTERS_WIDGET_NAME } from '@gspwidget/util';
import { WidgetEventMap } from './widget-event.decorator';
import { WidgetRegistry } from './widget-registry';
import * as i0 from "@angular/core";
export class WidgetEventService {
    constructor() { }
    /**
     * @param {?} widgetName
     * @param {?} pvs
     * @return {?}
     */
    getEventsByWidgetName(widgetName, pvs) {
        /** @type {?} */
        const ctorName = WidgetRegistry.getWidgetByName(widgetName).ctor.name;
        /** @type {?} */
        const widgetEvents = WidgetEventMap.get(ctorName) || [];
        if (widgetName === FILTERS_WIDGET_NAME) {
            return pvs[FILTERS_WIDGET_FILTERS_PROP_NAME].map((/**
             * @param {?} fc
             * @return {?}
             */
            (fc) => ({
                name: fc.i18nName[curLang],
                propName: fc.i18nName[curLang],
            })));
        }
        // 组装
        return [
            ...widgetEvents,
            ...test,
        ];
    }
}
WidgetEventService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
WidgetEventService.ctorParameters = () => [];
/** @nocollapse */ WidgetEventService.ngInjectableDef = i0.defineInjectable({ factory: function WidgetEventService_Factory() { return new WidgetEventService(); }, token: WidgetEventService, providedIn: "root" });
/** @type {?} */
const test = [{
        name: 'click bar'
    }, {
        name: 'event 2'
    }, {
        name: 'event 3'
    }];
