// @farris/ui-wizard/esm5/lib/page/page.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, Output, EventEmitter } from '@angular/core';
import { WizardComponent } from '../wizard/wizard.component';
import { PageService } from '../service/page.service';
var FarrisWizardPageComponent = /** @class */ (function () {
    function FarrisWizardPageComponent(wizardSet, pageSer) {
        this.pageSer = pageSer;
        // 唯一标记
        this.pageId = '';
        // 默认class
        this.addClass = true;
        this._active = false;
        this.pageSelected = new EventEmitter();
        this.wizardSet = wizardSet;
        this.wizardSet.addPage(this);
    }
    Object.defineProperty(FarrisWizardPageComponent.prototype, "activeCls", {
        get: /**
         * @return {?}
         */
        function () {
            return this._active;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FarrisWizardPageComponent.prototype, "active", {
        get: /**
         * @return {?}
         */
        function () {
            return this._active;
        },
        // 当前选中
        set: 
        // 当前选中
        /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._active = value;
            if (this.pageId) {
                this.doAfterActive();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FarrisWizardPageComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        //console.log('FarrisWizardPageComponent——' + this.pageId);
        this.pageSer.setPageId(this.pageId);
    };
    /**
     * @return {?}
     */
    FarrisWizardPageComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.doAfterActive();
    };
    /**
     * @return {?}
     */
    FarrisWizardPageComponent.prototype.getPageId = /**
     * @return {?}
     */
    function () {
        return this.pageId;
    };
    /**
     *
     * @param value
     */
    /**
     *
     * @param {?} value
     * @return {?}
     */
    FarrisWizardPageComponent.prototype.setPageActive = /**
     *
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this.active !== value) {
            this.active = value;
        }
    };
    /**
     * 处理Active
     */
    /**
     * 处理Active
     * @return {?}
     */
    FarrisWizardPageComponent.prototype.doAfterActive = /**
     * 处理Active
     * @return {?}
     */
    function () {
        this.pageSelected.emit({
            pageId: this.pageId,
            state: this.active
        });
        //console.log('FarrisWizardPageComponent——' + this.pageid + this.active);
        this.pageSer.activeState.next({
            pageId: this.pageId,
            state: this.active
        });
    };
    FarrisWizardPageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wizard-page',
                    template: '<ng-content></ng-content>',
                    providers: [PageService],
                    styles: [":host.f-component-wizard-page{display:none;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;-webkit-box-flex:1;flex-grow:1;flex-shrink:1;flex-basis:0;position:relative}:host.f-component-wizard-page.active{display:-webkit-box;display:flex}"]
                }] }
    ];
    /** @nocollapse */
    FarrisWizardPageComponent.ctorParameters = function () { return [
        { type: WizardComponent },
        { type: PageService }
    ]; };
    FarrisWizardPageComponent.propDecorators = {
        pageId: [{ type: Input }],
        addClass: [{ type: HostBinding, args: ['class.f-component-wizard-page',] }],
        activeCls: [{ type: HostBinding, args: ['class.active',] }],
        active: [{ type: Input }],
        pageSelected: [{ type: Output }]
    };
    return FarrisWizardPageComponent;
}());
export { FarrisWizardPageComponent };
if (false) {
    /** @type {?} */
    FarrisWizardPageComponent.prototype.pageId;
    /** @type {?} */
    FarrisWizardPageComponent.prototype.addClass;
    /**
     * @type {?}
     * @private
     */
    FarrisWizardPageComponent.prototype._active;
    /** @type {?} */
    FarrisWizardPageComponent.prototype.pageSelected;
    /** @type {?} */
    FarrisWizardPageComponent.prototype.wizardSet;
    /**
     * @type {?}
     * @private
     */
    FarrisWizardPageComponent.prototype.pageSer;
}
