// @farris/ui-forms/esm5/lib/radio/radio-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { IdService } from '@farris/ui-common';
var RadioGroupComponent = /** @class */ (function () {
    function RadioGroupComponent(idSer) {
        this.idSer = idSer;
        this.modelChange = new EventEmitter();
        // tslint:disable-next-line:ban-types
        this.subscriber = [];
        this.controlChange = (/**
         * @return {?}
         */
        function () { });
        this.controlTouch = (/**
         * @return {?}
         */
        function () { });
    }
    /**
     * @return {?}
     */
    RadioGroupComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.name = this.idSer.generate().replace(/-/g, '_');
    };
    /**
     * @param {?} value
     * @return {?}
     */
    RadioGroupComponent.prototype.changeModel = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.model = value;
        this.modelChange.emit(this.model);
        this.controlChange(value);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    RadioGroupComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.model = value;
        this.changeModel(this.model);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    RadioGroupComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.controlChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    RadioGroupComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.controlTouch = fn;
    };
    RadioGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-radio-group',
                    template: "\n    <div class=\"farris-radio-group btn-group\"\n    [class.farris-input-wrap]=\"type!=='button'\"\n    [class.btn-group-toggle]=\"type==='button'\"\n    [class.farris-checkradio-hor]=\"horizontal\">\n      <ng-content></ng-content>\n    </div>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return RadioGroupComponent; })),
                            multi: true,
                        }],
                    styles: ["\n    "]
                }] }
    ];
    /** @nocollapse */
    RadioGroupComponent.ctorParameters = function () { return [
        { type: IdService }
    ]; };
    RadioGroupComponent.propDecorators = {
        type: [{ type: Input }],
        model: [{ type: Input }],
        name: [{ type: Input }],
        horizontal: [{ type: Input }],
        modelChange: [{ type: Output }]
    };
    return RadioGroupComponent;
}());
export { RadioGroupComponent };
if (false) {
    /** @type {?} */
    RadioGroupComponent.prototype.type;
    /** @type {?} */
    RadioGroupComponent.prototype.model;
    /** @type {?} */
    RadioGroupComponent.prototype.name;
    /** @type {?} */
    RadioGroupComponent.prototype.horizontal;
    /** @type {?} */
    RadioGroupComponent.prototype.modelChange;
    /** @type {?} */
    RadioGroupComponent.prototype.subscriber;
    /**
     * @type {?}
     * @private
     */
    RadioGroupComponent.prototype.controlChange;
    /**
     * @type {?}
     * @private
     */
    RadioGroupComponent.prototype.controlTouch;
    /**
     * @type {?}
     * @private
     */
    RadioGroupComponent.prototype.idSer;
}
