// @gsp-sys/rtf-common/esm2015/lib/eventbus/frm-eventpipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// tslint:disable-next-line:import-blacklist
import { BehaviorSubject, Subscription, Subject } from 'rxjs';
import { Injectable } from '@angular/core';
export class EventBusPipe {
    /**
     * @param {?} tokenValue
     * @param {?} parentEventPipeList
     * @param {?} sub
     */
    constructor(tokenValue, parentEventPipeList, sub) {
        this.tokenValue = tokenValue;
        this.parentEventPipeList = parentEventPipeList;
        this.eventSubject = sub || new BehaviorSubject(null);
        this.subscription = new Subscription();
        this.onceSubscription = new Subscription();
        this.observers = new Map();
        if (this.parentEventPipeList) {
            this.parentEventPipeList.push(this);
        }
    }
    /**
     * @param {?} bizPara
     * @return {?}
     */
    post(bizPara) {
        this.eventSubject.next(bizPara);
    }
    /**
     * @param {?} eventHandler
     * @return {?}
     */
    subscribe(eventHandler) {
        this.subscription = this.eventSubject.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        (value) => {
            eventHandler(value);
        }));
        return this;
    }
    /**
     * @param {?} observerKey
     * @param {?} eventHandler
     * @return {?}
     */
    subscribeWithObkey(observerKey, eventHandler) {
        /** @type {?} */
        const observer = this.observers.get(observerKey);
        if (observer) {
            observer.unsubscribe();
        }
        /** @type {?} */
        const sub = this.eventSubject.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        (value) => {
            eventHandler(value);
        }));
        this.observers.set(observerKey, sub);
        return this;
    }
    /**
     * @param {?} eventHandler
     * @return {?}
     */
    subscribeOnce(eventHandler) {
        /** @type {?} */
        const self = this;
        /** @type {?} */
        const scription = this.eventSubject.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        (value) => {
            eventHandler(value);
            if (scription) {
                scription.unsubscribe();
            }
            self.parentEventPipeList = new Array();
        }));
        return this;
    }
    /**
     * @return {?}
     */
    unSubscribe() {
        if (this.subscription) {
            this.subscription.unsubscribe();
            this.subscription = null;
        }
    }
    // 注销使用once方法注册的订阅。
    /**
     * @return {?}
     */
    unSubscribeForOnce() {
        this.onceSubscription.unsubscribe();
    }
    /**
     * 取消某观察者的订阅
     * @param {?} observerKey 观察者唯一标识
     * @return {?}
     */
    unSubscribeWithObkey(observerKey) {
        /** @type {?} */
        const scription = this.observers.get(observerKey);
        if (scription) {
            scription.unsubscribe();
        }
    }
    /**
     * @param {?} tokenValue
     * @return {?}
     */
    matchEmitterToken(tokenValue) {
        if (this.tokenValue && tokenValue && this.tokenValue !== tokenValue) {
            return false;
        }
        return true;
    }
    /**
     * @param {?} tokenValue
     * @return {?}
     */
    examByTargetToken(tokenValue) {
        if (this.tokenValue !== tokenValue) {
            return false;
        }
        return true;
    }
    /**
     * @param {?} subscriber
     * @return {?}
     */
    dispose(subscriber) {
    }
}
EventBusPipe.decorators = [
    { type: Injectable }
];
/** @nocollapse */
EventBusPipe.ctorParameters = () => [
    { type: String },
    { type: Array },
    { type: Subject }
];
if (false) {
    /** @type {?} */
    EventBusPipe.prototype.eventSubject;
    /** @type {?} */
    EventBusPipe.prototype.observers;
    /** @type {?} */
    EventBusPipe.prototype.pipeCode;
    /**
     * @type {?}
     * @private
     */
    EventBusPipe.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    EventBusPipe.prototype.onceSubscription;
    /**
     * @type {?}
     * @private
     */
    EventBusPipe.prototype.tokenValue;
    /**
     * @type {?}
     * @private
     */
    EventBusPipe.prototype.parentEventPipeList;
}
