// @farris/lookup-config/esm5/lib/components/lookup-config-metadata.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/lookup-config-metadata.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input } from '@angular/core';
var LookupConfigMetaDataComponent = /** @class */ (function () {
    function LookupConfigMetaDataComponent() {
        // metadataPath = '';
        this.HelpMetadata = 'HelpMetadata';
        this.isDeisgnerEnv = false;
        this.selectedMeta = new EventEmitter();
        this.closeModal = new EventEmitter();
    }
    /**
     * @return {?}
     */
    LookupConfigMetaDataComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} event
     * @return {?}
     */
    LookupConfigMetaDataComponent.prototype.selectedMetadata = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.selectedMeta.emit(event);
    };
    /**
     * @return {?}
     */
    LookupConfigMetaDataComponent.prototype.closeModalEmitter = /**
     * @return {?}
     */
    function () {
        this.closeModal.emit();
    };
    LookupConfigMetaDataComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-lookup-config-metadata',
                    template: "<div class=\"f-lookup-config-metadata\">\r\n    <div class=\"f-lookup-config-metadata-inner\">\r\n        <app-select-metadatart typeList=\"HelpMetadata\" (selectedMetadata)=\"selectedMetadata($event)\"\r\n            (closeModalEmitter)=\"closeModalEmitter()\">\r\n        </app-select-metadatart>\r\n    </div> \r\n</div>"
                }] }
    ];
    /** @nocollapse */
    LookupConfigMetaDataComponent.ctorParameters = function () { return []; };
    LookupConfigMetaDataComponent.propDecorators = {
        metadataPath: [{ type: Input }],
        isDeisgnerEnv: [{ type: Input }],
        selectedMeta: [{ type: Output, args: ['selectedMeta',] }],
        closeModal: [{ type: Output, args: ['closeModal',] }]
    };
    return LookupConfigMetaDataComponent;
}());
export { LookupConfigMetaDataComponent };
if (false) {
    /** @type {?} */
    LookupConfigMetaDataComponent.prototype.HelpMetadata;
    /** @type {?} */
    LookupConfigMetaDataComponent.prototype.metadataPath;
    /** @type {?} */
    LookupConfigMetaDataComponent.prototype.isDeisgnerEnv;
    /** @type {?} */
    LookupConfigMetaDataComponent.prototype.selectedMeta;
    /** @type {?} */
    LookupConfigMetaDataComponent.prototype.closeModal;
}
