// @farris/dynamic-control-group/esm5/lib/interfaces/dynamic-control.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function Validator() { }
if (false) {
    /** @type {?} */
    Validator.prototype.name;
    /** @type {?} */
    Validator.prototype.validator;
    /** @type {?} */
    Validator.prototype.message;
}
/**
 * @record
 */
export function IDynamicBaseControl() { }
if (false) {
    /** @type {?} */
    IDynamicBaseControl.prototype.labelCode;
    /** @type {?} */
    IDynamicBaseControl.prototype.name;
    /** @type {?} */
    IDynamicBaseControl.prototype.control;
    /** @type {?|undefined} */
    IDynamicBaseControl.prototype.value;
    /** @type {?|undefined} */
    IDynamicBaseControl.prototype.className;
    /** @type {?|undefined} */
    IDynamicBaseControl.prototype.validations;
    /** @type {?} */
    IDynamicBaseControl.prototype.groupName;
}
/**
 * @record
 */
export function IDynamicDateControl() { }
if (false) {
    /** @type {?} */
    IDynamicDateControl.prototype.control;
}
/**
 * @record
 */
export function IDynamicDropdownControl() { }
if (false) {
    /** @type {?} */
    IDynamicDropdownControl.prototype.control;
}
/**
 * @record
 */
export function IDyanmicNumberControl() { }
if (false) {
    /** @type {?} */
    IDyanmicNumberControl.prototype.control;
}
/**
 * 默认生成一个hidden input，用于存放需要返回但不显示出的值。
 * mapFields：
 * @record
 */
export function IDyanmicHelpControl() { }
if (false) {
    /** @type {?} */
    IDyanmicHelpControl.prototype.control;
}
/** @enum {string} */
var DYNAMIC_CONTROL_TYPE = {
    HELP: 'help',
    TEXT: 'text',
    DATE_RANGE: 'date-range',
    SINGLE_DATE: 'single-date',
    SINGLE_NUMBER: 'single-number',
    NUMBER_RANGE: 'number-range',
    BASE_DROPDOWN: 'base-dropdown',
};
export { DYNAMIC_CONTROL_TYPE };
