// @farris/command-services/esm2015/lib/util/http.js
// https://github.com/typeslick/status-code-enum
var HttpStatusType;
(function (HttpStatusType) {
    HttpStatusType[HttpStatusType["Info"] = 1] = "Info";
    HttpStatusType[HttpStatusType["Success"] = 2] = "Success";
    HttpStatusType[HttpStatusType["Redirect"] = 3] = "Redirect";
    HttpStatusType[HttpStatusType["ClientError"] = 4] = "ClientError";
    HttpStatusType[HttpStatusType["ServerError"] = 5] = "ServerError";
})(HttpStatusType || (HttpStatusType = {}));
/**
 * http通用类
 */
class HttpUtil {
    /**
     * 基于http状态码返回状态类型
     * @param httpResponseSatus http请求返回状态码
     * @returns null | HttpStatusType
     */
    static getHttpStatusType(httpResponseSatus) {
        if (!httpResponseSatus) {
            return null;
        }
        let statusType = null; // 默认值为空
        if (httpResponseSatus < 100 || httpResponseSatus > 600) {
            statusType = null; // 不合理状态码
        }
        else if (httpResponseSatus >= 100 && httpResponseSatus < 200) {
            statusType = HttpStatusType.Info;
        }
        else if (httpResponseSatus >= 200 && httpResponseSatus < 300) {
            statusType = HttpStatusType.Success;
        }
        else if (httpResponseSatus >= 300 && httpResponseSatus < 400) {
            statusType = HttpStatusType.Redirect;
        }
        else if (httpResponseSatus >= 400 && httpResponseSatus < 500) {
            statusType = HttpStatusType.ClientError;
        }
        else if (httpResponseSatus >= 500 && httpResponseSatus < 600) {
            statusType = HttpStatusType.ServerError;
        }
        return statusType;
    }
}
export { HttpUtil, HttpStatusType };
