// @qdp/ide-cmp/esm2015/lib/service/schema/schema.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { SchemaManagerService, LcpService } from '@qdp/common';
import { Subject } from 'rxjs/Subject';
export class SchemaService {
    /**
     * @param {?} schemaManager
     * @param {?} lcpService
     */
    constructor(schemaManager, lcpService) {
        this.schemaManager = schemaManager;
        this.lcpService = lcpService;
        this.schemaInfo = new Subject();
        this.colList = new Subject();
        this.pageInfoChange = new Subject();
        this.operationChange = new Subject();
        this.schemaManager.schemaInfoChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            this.schemaInfo.next(value);
        }));
        this.schemaManager.colList.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            this.colList.next(value);
        }));
        this.lcpService.pageInfoChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            this.pageInfoChange.next(value);
        }));
        this.schemaManager.operationChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            this.operationChange.next(value);
        }));
    }
    /**
     * @param {?} queryId
     * @param {?} userId
     * @param {?} organizationId
     * @param {?} queryRelativeUrl
     * @param {?=} groupType
     * @return {?}
     */
    getSchemaList(queryId, userId, organizationId, queryRelativeUrl, groupType) {
        if (groupType) {
            return this.schemaManager.getSchemaListByGroupType(queryId, userId, organizationId, groupType, queryRelativeUrl);
        }
        else {
            return this.schemaManager.getSchemaList(queryId, userId, organizationId, queryRelativeUrl);
        }
    }
    /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @param {?=} groupType
     * @param {?=} organizationId
     * @return {?}
     */
    getSchema(id, queryId, queryRelativeUrl, groupType, organizationId) {
        return this.schemaManager.getSchema(id, queryId, queryRelativeUrl, groupType, organizationId);
    }
    /**
     * @param {?} schema
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    saveSchema(schema, queryRelativeUrl) {
        return this.schemaManager.saveSchema(schema, queryRelativeUrl);
    }
    /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @param {?=} groupType
     * @param {?=} organizationId
     * @return {?}
     */
    deleteSchema(id, queryId, queryRelativeUrl, groupType, organizationId) {
        return this.schemaManager.deleteSchema(id, queryId, queryRelativeUrl, groupType, organizationId);
    }
    /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} dynamicCols
     * @param {?} queryRelativeUrl
     * @param {?=} sheme
     * @param {?=} frozenCols
     * @param {?=} groupType
     * @return {?}
     */
    getSpreadTemplate(id, queryId, dynamicCols, queryRelativeUrl, sheme, frozenCols, groupType) {
        return this.schemaManager.getSpreadTemplate(id, queryId, dynamicCols, queryRelativeUrl, sheme, frozenCols, groupType);
    }
    /**
     * @param {?} param
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    getData(param, queryRelativeUrl) {
        return this.lcpService.getData(param, queryRelativeUrl);
    }
    /**
     * @param {?} params
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    getCol(params, queryRelativeUrl) {
        return this.lcpService.getCol(params, queryRelativeUrl);
    }
    /**
     * @return {?}
     */
    getOperation() {
        return this.schemaManager.getOperation();
    }
}
SchemaService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
SchemaService.ctorParameters = () => [
    { type: SchemaManagerService },
    { type: LcpService }
];
if (false) {
    /** @type {?} */
    SchemaService.prototype.schemaInfo;
    /** @type {?} */
    SchemaService.prototype.colList;
    /** @type {?} */
    SchemaService.prototype.pageInfoChange;
    /** @type {?} */
    SchemaService.prototype.operationChange;
    /**
     * @type {?}
     * @private
     */
    SchemaService.prototype.schemaManager;
    /**
     * @type {?}
     * @private
     */
    SchemaService.prototype.lcpService;
}
