// @gsp-sys/rtf-apphelp/esm2015/lib/rtf-app-help/service/rtf-app.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { RtfUIWebapiService } from '../../common/rtf-ui-webapi.service';
export class RtfAppService {
    /**
     * @param {?} rtfCommon
     * @param {?} httpClient
     */
    constructor(rtfCommon, httpClient) {
        this.rtfCommon = rtfCommon;
        this.httpClient = httpClient;
    }
    /**
     * 获取功能菜单操作
     * @return {?}
     */
    getAppList() {
        /** @type {?} */
        const url = this.rtfCommon.functionsApi;
        /** @type {?} */
        const header = this.rtfCommon.httpHeader;
        return this.httpClient.get(url, header);
    }
    /**
     * 获取包含BO数据的应用数据
     * @return {?}
     */
    getAppListWithBizObjectData() {
        /** @type {?} */
        const url = this.rtfCommon.appsWithBOApi;
        /** @type {?} */
        const header = this.rtfCommon.httpHeader;
        return this.httpClient.get(url, header);
    }
    /**
     * 获取多语资源文件
     * @param {?} url
     * @return {?}
     */
    getI18nJson(url) {
        /** @type {?} */
        const header = this.rtfCommon.httpHeader;
        return this.httpClient.get(url, header);
    }
}
RtfAppService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RtfAppService.ctorParameters = () => [
    { type: RtfUIWebapiService },
    { type: HttpClient }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    RtfAppService.prototype.rtfCommon;
    /**
     * @type {?}
     * @private
     */
    RtfAppService.prototype.httpClient;
}
