// @farris/extend-page-modal/esm5/lib/page-sidebar.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/page-sidebar.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output, Renderer2, ElementRef, ViewChild, HostBinding, HostListener } from '@angular/core';
import { DomSanitizer } from "@angular/platform-browser";
import { FPageModalDISMISS_REASONS, FEPageModalOptions } from './page-modal-options.class';
var FEPageModalSidebarComponent = /** @class */ (function () {
    function FEPageModalSidebarComponent(options, el, _renderer, sanitizer) {
        this.el = el;
        this._renderer = _renderer;
        this.sanitizer = sanitizer;
        this.isShown = false;
        this.closed = new EventEmitter();
        this.opened = new EventEmitter();
        this.resized = new EventEmitter();
        this.modalID = "MODALID_" + Date.now().toPrecision();
        this.iframeSrc = '';
        this._iframe = '';
        this.isModalHiding = false;
        this.canRemoveInitNoAnimateCls = true;
        this.sidebarCls = {
            'parent': 'f-sidebar-parent',
            'openParent': 'f-sidebar-parent-expand',
            'closeParent': 'f-sidebar-parent-collapse',
            'init': 'animated-none'
        };
        this.isOpen = false;
        this.config = Object.assign({}, options);
    }
    Object.defineProperty(FEPageModalSidebarComponent.prototype, "iframe", {
        get: /**
         * @return {?}
         */
        function () {
            return this._iframe;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (this.sanitizer) {
                this._iframe = value;
                this.iframeSrc = this.sanitizer.bypassSecurityTrustResourceUrl(value);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    FEPageModalSidebarComponent.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var tag = (/** @type {?} */ (event.target));
        if (this.config.ignoreBackdropClick || this.config.backdrop === 'static' || event.target !== this.el.nativeElement) {
            event.stopPropagation();
            return;
        }
        this.bsModalService.setDismissReason(FPageModalDISMISS_REASONS.BACKRDOP);
        this.close();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    FEPageModalSidebarComponent.prototype.onEsc = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this.isShown) {
            return;
        }
        if (event.keyCode === 27) {
            event.preventDefault();
        }
        if (this.config.keyboard &&
            this.level === this.bsModalService.getModalsCount()) {
            this.bsModalService.setDismissReason(FPageModalDISMISS_REASONS.ESC);
            this.close(event);
        }
    };
    /**
     * @return {?}
     */
    FEPageModalSidebarComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.parentEl = this.el.nativeElement.parentElement;
        this._renderer.addClass(this.parentEl, this.sidebarCls.parent);
        this.removeNoneAnimateCls();
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.isShown = true;
            _this.isOpen = true;
            // this._renderer.addClass(
            //     this.el.nativeElement, FEPageModalCLASS_NAME.SHOW
            // );
            // Parent
            _this._renderer.removeClass(_this.parentEl, _this.sidebarCls.closeParent);
            _this._renderer.addClass(_this.parentEl, _this.sidebarCls.openParent);
            // this._renderer.setStyle(this.el.nativeElement, 'display', 'block');
            // if (this.config.fitContent && this.el.nativeElement.className.indexOf('f-modal-fitContent-scroll') > -1) {
            //   this.el.nativeElement.scrollTop = 0;
            // }
            // if (document && document.body) {
            //     const modalLen = this.bsModalService.getModalsCount();
            //     if (modalLen === 1) {
            //         this.bsModalService.checkScrollbar();
            //         this.bsModalService.setScrollbar();
            //     }
            //     if (!document.body.classList.contains(FEPageModalCLASS_NAME.OPEN)) {
            //         this._renderer.addClass(document.body, FEPageModalCLASS_NAME.OPEN);
            //     }
            // }
            _this.opened.emit({ modal: _this });
        }), 0);
        if (this.el.nativeElement) {
            this.el.nativeElement.focus();
        }
    };
    /**
     * @return {?}
     */
    FEPageModalSidebarComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    FEPageModalSidebarComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.isShown) {
            this.close();
        }
    };
    /**
     * @return {?}
     */
    FEPageModalSidebarComponent.prototype.hasIframe = /**
     * @return {?}
     */
    function () {
        return this._iframe != '';
    };
    /**
     * @return {?}
     */
    FEPageModalSidebarComponent.prototype.closeSideBar = /**
     * @return {?}
     */
    function () {
        // this.removeNoneAnimateCls();
        // if (this.beforeClose) {
        //     this.beforeClose().subscribe((isOpen: boolean) => {
        //         this.isOpen = !isOpen;
        //     })
        // } else {
        //     this.isOpen = false;
        // }
        this._renderer.removeClass(this.parentEl, this.sidebarCls.openParent);
        this._renderer.addClass(this.parentEl, this.sidebarCls.closeParent);
    };
    /**
     * @return {?}
     */
    FEPageModalSidebarComponent.prototype.openSideBar = /**
     * @return {?}
     */
    function () {
        this.removeNoneAnimateCls();
        this.isOpen = true;
        this._renderer.removeClass(this.parentEl, this.sidebarCls.closeParent);
        this._renderer.addClass(this.parentEl, this.sidebarCls.openParent);
    };
    /*侧边栏初始因为动画收起时，会被看到, 移除 */
    /*侧边栏初始因为动画收起时，会被看到, 移除 */
    /**
     * @private
     * @return {?}
     */
    FEPageModalSidebarComponent.prototype.removeNoneAnimateCls = /*侧边栏初始因为动画收起时，会被看到, 移除 */
    /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.canRemoveInitNoAnimateCls) {
            return;
        }
        if (this.sidebarMainEl) {
            this._renderer.removeClass(this.sidebarMainEl.nativeElement, this.sidebarCls.init);
            this.canRemoveInitNoAnimateCls = false;
        }
    };
    // 侧边栏宽度计算
    // 侧边栏宽度计算
    /**
     * @return {?}
     */
    FEPageModalSidebarComponent.prototype.getSidebarWidth = 
    // 侧边栏宽度计算
    /**
     * @return {?}
     */
    function () {
        if (typeof this.config['width'] == 'string') {
            return this.config['width'];
        }
        else {
            return this.config['width'] + 'px';
        }
    };
    // 必须带的方法
    // 必须带的方法
    /**
     * @param {?=} e
     * @param {?=} isCloseButton
     * @return {?}
     */
    FEPageModalSidebarComponent.prototype.close = 
    // 必须带的方法
    /**
     * @param {?=} e
     * @param {?=} isCloseButton
     * @return {?}
     */
    function (e, isCloseButton) {
        var _this = this;
        if (isCloseButton === void 0) { isCloseButton = false; }
        if (e) {
            e.stopPropagation();
        }
        if (this.isModalHiding || !this.isShown) {
            return;
        }
        this.config.beforeClose(this).subscribe((/**
         * @param {?} r
         * @return {?}
         */
        function (r) {
            if (!r) {
                return false;
            }
            //动画结束执行其他操作
            _this.sidebarMainEl.nativeElement.parentElement.addEventListener('transitionend', (/**
             * @param {?} e
             * @return {?}
             */
            function (e) {
                _this.bsModalService.hide(_this.level);
                _this.isModalHiding = false;
                if (_this.config.closed) {
                    _this.config.closed(isCloseButton);
                }
                _this.closed.emit(_this.level);
            }), false);
            _this.isModalHiding = true;
            _this.isShown = false;
            _this._renderer.removeClass(_this.parentEl, _this.sidebarCls.openParent);
            _this._renderer.addClass(_this.parentEl, _this.sidebarCls.closeParent);
        }));
    };
    /**
     * @return {?}
     */
    FEPageModalSidebarComponent.prototype.toCenter = /**
     * @return {?}
     */
    function () {
    };
    FEPageModalSidebarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fe-page-sidebar',
                    template: "<div [attr.id]=\"modalID\" class=\"f-cmp-sidebar f-sidebar-slide fe-page-sidebar\" [class]=\"config.class+' '+'f-sidebar-pos-'+config.sidebar['showPos']\"\r\n  [ngClass]=\"{'f-sidebar-show': isShown||isOpen, 'f-sidebar-collapse': !isShown||!isOpen}\">\r\n  <div class=\"f-sidebar-main animated-none\" [ngClass]=\"{'f-sidebar-main--has-actions':config.showCloseButton}\" #sidebarMain [ngStyle]=\"{width:config.width}\"\r\n    [f-area-response]=\"!iframe&&config.areaResponse\">\r\n    <perfect-scrollbar [config]=\"{}\" [disabled]=\"!config.enableScroll||config.fitContent\" *ngIf=\"!iframe\">\r\n      <ng-content></ng-content>\r\n    </perfect-scrollbar>\r\n    <iframe *ngIf=\"iframe\" class=\"f-utils-fill\" width=\"100%\" frameborder=\"0\" [src]=\"iframeSrc\"></iframe>\r\n    <div class=\"f-sidebar--action-list\" *ngIf=\"config.showCloseButton\">\r\n      <div class=\"action-list--item\" (click)=\"close($event,true)\">\r\n        <span class=\"action-list--icon close-icon\">\r\n          <i class=\"f-icon f-icon-close\"></i>\r\n        </span>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</div>",
                    host: {
                        class: 'fe-cmp-page-modal'
                    }
                }] }
    ];
    /** @nocollapse */
    FEPageModalSidebarComponent.ctorParameters = function () { return [
        { type: FEPageModalOptions },
        { type: ElementRef },
        { type: Renderer2 },
        { type: DomSanitizer }
    ]; };
    FEPageModalSidebarComponent.propDecorators = {
        closed: [{ type: Output }],
        opened: [{ type: Output }],
        resized: [{ type: Output }],
        modalID: [{ type: HostBinding, args: ['id',] }],
        iframe: [{ type: Input }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }],
        onEsc: [{ type: HostListener, args: ['window:keydown.esc', ['$event'],] }],
        sidebarMainEl: [{ type: ViewChild, args: ['sidebarMain',] }]
    };
    return FEPageModalSidebarComponent;
}());
export { FEPageModalSidebarComponent };
if (false) {
    /** @type {?} */
    FEPageModalSidebarComponent.prototype.config;
    /** @type {?} */
    FEPageModalSidebarComponent.prototype.isShown;
    /** @type {?} */
    FEPageModalSidebarComponent.prototype.level;
    /** @type {?} */
    FEPageModalSidebarComponent.prototype.isAnimated;
    /** @type {?} */
    FEPageModalSidebarComponent.prototype.bsModalService;
    /** @type {?} */
    FEPageModalSidebarComponent.prototype.dialogType;
    /** @type {?} */
    FEPageModalSidebarComponent.prototype.closed;
    /** @type {?} */
    FEPageModalSidebarComponent.prototype.opened;
    /** @type {?} */
    FEPageModalSidebarComponent.prototype.resized;
    /** @type {?} */
    FEPageModalSidebarComponent.prototype.modalID;
    /** @type {?} */
    FEPageModalSidebarComponent.prototype.iframeSrc;
    /**
     * @type {?}
     * @private
     */
    FEPageModalSidebarComponent.prototype._iframe;
    /**
     * @type {?}
     * @private
     */
    FEPageModalSidebarComponent.prototype.isModalHiding;
    /**
     * @type {?}
     * @private
     */
    FEPageModalSidebarComponent.prototype.sidebarMainEl;
    /**
     * @type {?}
     * @private
     */
    FEPageModalSidebarComponent.prototype.canRemoveInitNoAnimateCls;
    /**
     * @type {?}
     * @private
     */
    FEPageModalSidebarComponent.prototype.parentEl;
    /**
     * @type {?}
     * @private
     */
    FEPageModalSidebarComponent.prototype.sidebarCls;
    /** @type {?} */
    FEPageModalSidebarComponent.prototype.isOpen;
    /**
     * @type {?}
     * @protected
     */
    FEPageModalSidebarComponent.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FEPageModalSidebarComponent.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    FEPageModalSidebarComponent.prototype.sanitizer;
}
