// @farris/extend-page-modal/esm5/lib/page-modal-options.class.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/page-modal-options.class.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { of } from 'rxjs';
var FEPageModalOptions = /** @class */ (function () {
    function FEPageModalOptions() {
    }
    FEPageModalOptions.decorators = [
        { type: Injectable }
    ];
    return FEPageModalOptions;
}());
export { FEPageModalOptions };
if (false) {
    /**
     *  Includes a modal-backdrop element. Alternatively,
     *  specify static for a backdrop which doesn't close the modal on click.
     * @type {?}
     */
    FEPageModalOptions.prototype.backdrop;
    /**
     * Closes the modal when escape key is pressed.
     * @type {?}
     */
    FEPageModalOptions.prototype.keyboard;
    /** @type {?} */
    FEPageModalOptions.prototype.focus;
    /**
     * Shows the modal when initialized.
     * @type {?}
     */
    FEPageModalOptions.prototype.show;
    /**
     * Ignore the backdrop click
     * @type {?}
     */
    FEPageModalOptions.prototype.ignoreBackdropClick;
    /**
     * Css class for opened modal
     * @type {?}
     */
    FEPageModalOptions.prototype.class;
    /**
     * Toggle animation
     * @type {?}
     */
    FEPageModalOptions.prototype.animated;
    /**
     * Modal data
     * @type {?}
     */
    FEPageModalOptions.prototype.initialState;
    /**
     * 显示关闭按钮
     * @type {?}
     */
    FEPageModalOptions.prototype.showCloseButton;
    /**
     * 显示最大化按钮
     * @type {?}
     */
    FEPageModalOptions.prototype.showMaxButton;
    /**
     * 显示最小化按钮
     * @type {?}
     */
    FEPageModalOptions.prototype.showMinButton;
    /**
     * 窗口宽度
     * @type {?}
     */
    FEPageModalOptions.prototype.width;
    /**
     * 窗口高度
     * @type {?}
     */
    FEPageModalOptions.prototype.height;
    /** @type {?} */
    FEPageModalOptions.prototype.minWidth;
    /** @type {?} */
    FEPageModalOptions.prototype.minHeight;
    /** @type {?} */
    FEPageModalOptions.prototype.maxWidth;
    /** @type {?} */
    FEPageModalOptions.prototype.maxHeight;
    /**
     * 启用拖拽功能
     * @type {?}
     */
    FEPageModalOptions.prototype.draggable;
    /**
     * 启用调整尺寸大小
     * @type {?}
     */
    FEPageModalOptions.prototype.resizable;
    /**
     * 是否启用滚动条
     * @type {?}
     */
    FEPageModalOptions.prototype.enableScroll;
    /** @type {?} */
    FEPageModalOptions.prototype.beforeOpen;
    /** @type {?} */
    FEPageModalOptions.prototype.beforeClose;
    /** @type {?} */
    FEPageModalOptions.prototype.opened;
    /** @type {?} */
    FEPageModalOptions.prototype.closed;
    /** @type {?} */
    FEPageModalOptions.prototype.okText;
    /** @type {?} */
    FEPageModalOptions.prototype.cancelText;
    /**
     * 弹窗类型：default，弹出组件、模板等；iframe,弹出iframe; ngFactory, 弹出ngFactory; sidebar:侧边栏
     * @type {?}
     */
    FEPageModalOptions.prototype.dialogType;
    /** @type {?} */
    FEPageModalOptions.prototype.fitContent;
    /** @type {?} */
    FEPageModalOptions.prototype.areaResponse;
    /** @type {?} */
    FEPageModalOptions.prototype.sidebar;
}
/** @type {?} */
export var FEPModalConfigDefaults = {
    backdrop: 'static',
    keyboard: true,
    focus: true,
    show: false,
    ignoreBackdropClick: true,
    class: '',
    animated: true,
    initialState: {},
    width: 600,
    height: 450,
    showCloseButton: true,
    minHeight: null,
    minWidth: null,
    maxHeight: null,
    maxWidth: null,
    draggable: true,
    resizable: true,
    enableScroll: true,
    beforeOpen: (/**
     * @return {?}
     */
    function () { return of(true); }),
    beforeClose: (/**
     * @return {?}
     */
    function () { return of(true); }),
    dialogType: 'modal',
    fitContent: false,
    areaResponse: true,
    sidebar: {
        showPos: 'right',
        parentCls: ''
    }
};
/** @type {?} */
export var FEPageModalCLASS_NAME = {
    SCROLLBAR_MEASURER: 'modal-scrollbar-measure',
    BACKDROP: 'modal-backdrop',
    OPEN: 'modal-open',
    FADE: 'fade',
    IN: 'in',
    // bs3
    SHOW: 'show' // bs4
};
/** @type {?} */
export var FPageModalSELECTOR = {
    DIALOG: '.modal-dialog',
    DATA_TOGGLE: '[data-toggle="modal"]',
    DATA_DISMISS: '[data-dismiss="modal"]',
    FIXED_CONTENT: '.navbar-fixed-top, .navbar-fixed-bottom, .is-fixed'
};
/** @type {?} */
export var FPageModalTRANSITION_DURATIONS = {
    MODAL: 300,
    BACKDROP: 150
};
/** @type {?} */
export var FPageModalDISMISS_REASONS = {
    BACKRDOP: 'backdrop-click',
    ESC: 'esc'
};
