// @farris/devkit/esm2015/lib/effector/dependency_effector.js
import { Inject, Injectable, Injector } from '@angular/core';
import { BindingData } from '../binding-data/index';
import { EntityList } from '../entity/index';
import { NAMESPACE } from '../frame/tokens';
import { Repository } from '../repository/index';
/**
 * 依赖副作用器
 * @description 当表达式计算结果为true时，清空字段值
 * 当value==true且属性当前值不是null时，设置属性值为null
 */
export class DependencyEffector {
    constructor(injector, repository, namespace, bindingData) {
        this.injector = injector;
        this.repository = repository;
        this.namespace = namespace;
        this.bindingData = bindingData;
        this.ns = namespace;
    }
    effect(path, value, options) {
        if (!options || !options.path) {
            throw new Error('DependencyEffector 需要指定行信息。');
        }
        if (typeof value !== 'boolean') {
            console.warn(`DependencyEffector 依赖表达式计算结果应该为true/false，当前值为：${value}`);
        }
        const propertyPath = options.path;
        const rowId = propertyPath[0] || this.bindingData.list.currentItem.primaryKeyValue;
        let entity = this.repository.entityCollection.getEntityById(rowId);
        if (rowId && !entity) {
            throw new Error(`找不到id：${rowId}对应的实体！`);
        }
        // propertyPath like : [1, child1s, 1.1, child2s, 1.1.1,propName] or [1,udt,propName] or [1,prop]
        const propName = propertyPath.pop();
        let object = entity;
        for (let index = 1; index < propertyPath.length; index++) {
            const propertyName = propertyPath[index];
            if (object instanceof EntityList) {
                object = object.get(propertyName);
            }
            else {
                object = object[propertyName];
            }
        }
        if (object) {
            if (object[propName] !== null && value === true) {
                object[propName] = null;
            }
        }
        else {
            throw new Error(`[DependencyEffector] 找不到实体对应的路径：${propertyPath.push(propName)}`);
        }
    }
}
DependencyEffector.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DependencyEffector.ctorParameters = () => [
    { type: Injector },
    { type: Repository },
    { type: undefined, decorators: [{ type: Inject, args: [NAMESPACE,] }] },
    { type: BindingData }
];
