// @gsp-cmp/webcommand/esm2015/lib/vm/BranchCommandItem.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommandItemType } from "./CommandItemType";
import { CommandItemConvertor } from "./ICommandItem";
export class BranchCommandItem {
    /**
     * @return {?}
     */
    GetItemId() {
        return this.Id;
    }
    /**
     * @return {?}
     */
    GetItemCode() {
        return this.Code;
    }
    /**
     * @return {?}
     */
    GetItemName() {
        return this.Name;
    }
    /**
     * @return {?}
     */
    GetItemType() {
        return CommandItemType.Branch;
    }
}
if (false) {
    /** @type {?} */
    BranchCommandItem.prototype.Id;
    /** @type {?} */
    BranchCommandItem.prototype.Code;
    /** @type {?} */
    BranchCommandItem.prototype.Name;
    /** @type {?} */
    BranchCommandItem.prototype.ConditionType;
    /** @type {?} */
    BranchCommandItem.prototype.Express;
    /** @type {?} */
    BranchCommandItem.prototype.Items;
}
export class BranchCommandItemConvertor {
    /**
     * @param {?} obj
     * @return {?}
     */
    ConvertJObject(obj) {
        /** @type {?} */
        let branchItem = (/** @type {?} */ (obj));
        /** @type {?} */
        let jobj = new Object();
        jobj["Id"] = branchItem.Id;
        jobj["Code"] = branchItem.Code;
        jobj["Name"] = branchItem.Name;
        jobj["ConditionType"] = branchItem.ConditionType;
        jobj["Express"] = branchItem.Express;
        if (branchItem.Items != null) {
            /** @type {?} */
            let itemArray = [];
            /** @type {?} */
            let itemConvertor = new CommandItemConvertor();
            branchItem.Items.forEach((/**
             * @param {?} element
             * @return {?}
             */
            element => {
                itemArray.push(itemConvertor.ConvertJObject(element));
            }));
            jobj["Items"] = itemArray;
        }
        return jobj;
    }
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    InitFromJobject(jsonObj) {
        /** @type {?} */
        let branchItem = new BranchCommandItem();
        branchItem.Id = jsonObj["Id"];
        branchItem.Code = jsonObj["Code"];
        branchItem.Name = jsonObj["Name"];
        branchItem.ConditionType = jsonObj["ConditionType"];
        branchItem.Express = jsonObj["Express"];
        if (jsonObj["Items"] != null) {
            branchItem.Items = new Array();
            /** @type {?} */
            let itemConvertor = new CommandItemConvertor();
            jsonObj["Items"].forEach((/**
             * @param {?} element
             * @return {?}
             */
            element => {
                branchItem.Items.push(itemConvertor.InitFromJobject(element));
            }));
        }
        return branchItem;
    }
}
