// @gsp-cmp/webcommand/esm2015/lib/vm/CmpMethodParamConfig.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class CmpMethodParamConfig {
}
if (false) {
    /** @type {?} */
    CmpMethodParamConfig.prototype.ParamCode;
    /** @type {?} */
    CmpMethodParamConfig.prototype.ParamName;
    /** @type {?} */
    CmpMethodParamConfig.prototype.ParamExpress;
}
export class CmpMethodParamConfigConvert {
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    InitFromJobject(jsonObj) {
        /** @type {?} */
        let cmpMethodParamConfig = new CmpMethodParamConfig();
        cmpMethodParamConfig.ParamCode = jsonObj["ParamCode"];
        cmpMethodParamConfig.ParamName = jsonObj["ParamName"];
        cmpMethodParamConfig.ParamExpress = jsonObj["ParamExpress"];
        return cmpMethodParamConfig;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    ConvertJObject(obj) {
        /** @type {?} */
        let cmpMethodParamConfig = (/** @type {?} */ (obj));
        /** @type {?} */
        let jobj = new Object();
        jobj["ParamCode"] = cmpMethodParamConfig.ParamCode;
        jobj["ParamName"] = cmpMethodParamConfig.ParamName;
        jobj["ParamExpress"] = cmpMethodParamConfig.ParamExpress;
        return jobj;
    }
}
