// @gsp-cmp/webcommand/esm2015/lib/vm/CmpMethodRefering.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CmpMethodParamConfigConvert } from "./CmpMethodParamConfig";
import { CommandItemType } from "./CommandItemType";
export class CmpMethodRefering {
    /**
     * @return {?}
     */
    GetItemType() {
        return CommandItemType.MethodRefer;
    }
    /**
     * @return {?}
     */
    GetItemCode() {
        return this.Code;
    }
    /**
     * @return {?}
     */
    GetItemName() {
        return this.Name;
    }
    /**
     * @return {?}
     */
    GetItemId() {
        return this.Id;
    }
}
if (false) {
    /** @type {?} */
    CmpMethodRefering.prototype.Id;
    /** @type {?} */
    CmpMethodRefering.prototype.Code;
    /** @type {?} */
    CmpMethodRefering.prototype.Name;
    /** @type {?} */
    CmpMethodRefering.prototype.ComponentId;
    /** @type {?} */
    CmpMethodRefering.prototype.ComponentCode;
    /** @type {?} */
    CmpMethodRefering.prototype.ComponentName;
    /** @type {?} */
    CmpMethodRefering.prototype.ComponentPath;
    /** @type {?} */
    CmpMethodRefering.prototype.MethodId;
    /** @type {?} */
    CmpMethodRefering.prototype.MethodCode;
    /** @type {?} */
    CmpMethodRefering.prototype.MethodName;
    /** @type {?} */
    CmpMethodRefering.prototype.IsReplaced;
    /** @type {?} */
    CmpMethodRefering.prototype.IsBeforeExpansion;
    /** @type {?} */
    CmpMethodRefering.prototype.IsAfterExpansion;
    /** @type {?} */
    CmpMethodRefering.prototype.ParamConfigs;
}
export class CmpMethodReferingConvert {
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    InitFromJobject(jsonObj) {
        /** @type {?} */
        let cmpMethodRefering = new CmpMethodRefering();
        cmpMethodRefering.Id = jsonObj["Id"];
        cmpMethodRefering.Code = jsonObj["Code"];
        cmpMethodRefering.Name = jsonObj["Name"];
        cmpMethodRefering.MethodId = jsonObj["MethodId"];
        cmpMethodRefering.MethodCode = jsonObj["MethodCode"];
        cmpMethodRefering.MethodName = jsonObj["MethodName"];
        cmpMethodRefering.ComponentId = jsonObj["ComponentId"];
        cmpMethodRefering.ComponentCode = jsonObj["ComponentCode"];
        cmpMethodRefering.ComponentName = jsonObj["ComponentName"];
        cmpMethodRefering.ComponentPath = jsonObj["ComponentPath"];
        cmpMethodRefering.IsReplaced = jsonObj["IsReplaced"];
        cmpMethodRefering.IsBeforeExpansion = jsonObj["IsBeforeExpansion"];
        cmpMethodRefering.IsAfterExpansion = jsonObj["IsAfterExpansion"];
        if (jsonObj["ParamConfigs"] != null) {
            cmpMethodRefering.ParamConfigs = new Array();
            /** @type {?} */
            let convertor = new CmpMethodParamConfigConvert();
            jsonObj["ParamConfigs"].forEach((/**
             * @param {?} element
             * @return {?}
             */
            element => {
                cmpMethodRefering.ParamConfigs.push(convertor.InitFromJobject(element));
            }));
        }
        return cmpMethodRefering;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    ConvertJObject(obj) {
        /** @type {?} */
        let cmpMethodRefering = (/** @type {?} */ (obj));
        /** @type {?} */
        let jobj = new Object();
        jobj["Id"] = cmpMethodRefering.Id;
        jobj["Code"] = cmpMethodRefering.Code;
        jobj["Name"] = cmpMethodRefering.Name;
        jobj["MethodId"] = cmpMethodRefering.MethodId;
        jobj["MethodCode"] = cmpMethodRefering.MethodCode;
        jobj["MethodName"] = cmpMethodRefering.MethodName;
        jobj["ComponentId"] = cmpMethodRefering.ComponentId;
        jobj["ComponentCode"] = cmpMethodRefering.ComponentCode;
        jobj["ComponentName"] = cmpMethodRefering.ComponentName;
        jobj["ComponentPath"] = cmpMethodRefering.ComponentPath;
        jobj["IsReplaced"] = cmpMethodRefering.IsReplaced;
        jobj["IsBeforeExpansion"] = cmpMethodRefering.IsBeforeExpansion;
        jobj["IsAfterExpansion"] = cmpMethodRefering.IsAfterExpansion;
        if (cmpMethodRefering.ParamConfigs != null) {
            /** @type {?} */
            let CommandsJArry = jobj["ParamConfigs"] = [];
            /** @type {?} */
            let convertor = new CmpMethodParamConfigConvert();
            cmpMethodRefering.ParamConfigs.forEach((/**
             * @param {?} paramConfig
             * @return {?}
             */
            paramConfig => {
                CommandsJArry.push(convertor.ConvertJObject(paramConfig));
            }));
        }
        return jobj;
    }
}
