// @gsp-cmp/webcommand/esm2015/lib/vm/Command.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommandItemConvertor } from "./ICommandItem";
export class WebCommand {
}
if (false) {
    /** @type {?} */
    WebCommand.prototype.Id;
    /** @type {?} */
    WebCommand.prototype.Code;
    /** @type {?} */
    WebCommand.prototype.Name;
    /** @type {?} */
    WebCommand.prototype.Description;
    /** @type {?} */
    WebCommand.prototype.Parameters;
    /** @type {?} */
    WebCommand.prototype.SourceCode;
    /** @type {?} */
    WebCommand.prototype.Items;
}
export class WebCommandConvertor {
    /**
     * @param {?} obj
     * @return {?}
     */
    ConvertJObject(obj) {
        /** @type {?} */
        let cmpOp = (/** @type {?} */ (obj));
        /** @type {?} */
        let jobj = new Object();
        jobj["Id"] = cmpOp.Id;
        jobj["Code"] = cmpOp.Code;
        jobj["Name"] = cmpOp.Name;
        jobj["Description"] = cmpOp.Description;
        jobj["SourceCode"] = cmpOp.SourceCode;
        jobj["Parameters"] = cmpOp.Parameters;
        /** @type {?} */
        let items = [];
        if (cmpOp.Items != null) {
            /** @type {?} */
            let itemConvertor = new CommandItemConvertor();
            cmpOp.Items.forEach((/**
             * @param {?} element
             * @return {?}
             */
            element => {
                items.push(itemConvertor.ConvertJObject(element));
            }));
        }
        jobj["Items"] = items;
        return jobj;
    }
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    InitFromJobject(jsonObj) {
        /** @type {?} */
        let cmpOp = new WebCommand();
        cmpOp.Id = jsonObj["Id"];
        cmpOp.Code = jsonObj["Code"];
        cmpOp.Name = jsonObj["Name"];
        cmpOp.Description = jsonObj["Description"];
        cmpOp.SourceCode = jsonObj["SourceCode"];
        cmpOp.Parameters = jsonObj["Parameters"];
        cmpOp.Items = new Array();
        if (jsonObj["Items"] != null) {
            /** @type {?} */
            let itemConvertor = new CommandItemConvertor();
            jsonObj["Items"].forEach((/**
             * @param {?} element
             * @return {?}
             */
            element => {
                cmpOp.Items.push(itemConvertor.InitFromJobject(element));
            }));
        }
        return cmpOp;
    }
}
