// @gsp-cmp/webcommand/esm2015/lib/vm/ExtendProperty.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class ExtendProperty {
}
if (false) {
    /** @type {?} */
    ExtendProperty.prototype.FormCode;
    /** @type {?} */
    ExtendProperty.prototype.IsCommon;
}
export class ExtendsConvert {
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    InitFromJobject(jsonObj) {
        if (jsonObj) {
            /** @type {?} */
            let extendProperty = new ExtendProperty();
            extendProperty.FormCode = jsonObj["FormCode"];
            extendProperty.IsCommon = jsonObj["IsCommon"];
            return extendProperty;
        }
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    ConvertJObject(obj) {
        /** @type {?} */
        let extendProperty = (/** @type {?} */ (obj));
        /** @type {?} */
        let jobj = new Object();
        jobj["FormCode"] = extendProperty.FormCode;
        jobj["IsCommon"] = extendProperty.IsCommon;
        return jobj;
    }
}
