// @gsp-cmp/webcommand/esm2015/lib/vm/ICommandItem.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommandItemType } from "./CommandItemType";
import { CmpMethodRefering } from "./CmpMethodRefering";
import { BranchCommandItemConvertor } from "./BranchCommandItem";
import { BranchCollectionCommandItemConvertor } from "./BranchCollectionCommandItem";
/**
 * @record
 */
export function ICommandItem() { }
if (false) {
    /**
     * @return {?}
     */
    ICommandItem.prototype.GetItemType = function () { };
    /**
     * @return {?}
     */
    ICommandItem.prototype.GetItemCode = function () { };
    /**
     * @return {?}
     */
    ICommandItem.prototype.GetItemName = function () { };
    /**
     * @return {?}
     */
    ICommandItem.prototype.GetItemId = function () { };
}
export class CommandItemConvertor {
    /**
     * @param {?} obj
     * @return {?}
     */
    ConvertJObject(obj) {
        /** @type {?} */
        let commandItem = (/** @type {?} */ (obj));
        /** @type {?} */
        let itemType = commandItem.GetItemType();
        /** @type {?} */
        let jobj = new Object();
        jobj["Type"] = itemType;
        if (itemType == CommandItemType.MethodRefer) {
            jobj["Content"] = commandItem;
        }
        else if (itemType == CommandItemType.Branch) {
            /** @type {?} */
            let branchConvertor = new BranchCommandItemConvertor();
            jobj["Content"] = branchConvertor.ConvertJObject((/** @type {?} */ (commandItem)));
        }
        else if (itemType == CommandItemType.BranchCollection) {
            /** @type {?} */
            let branchCollectionConvertor = new BranchCollectionCommandItemConvertor();
            jobj["Content"] = branchCollectionConvertor.ConvertJObject((/** @type {?} */ (commandItem)));
        }
        return jobj;
    }
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    InitFromJobject(jsonObj) {
        /** @type {?} */
        let itemType = (/** @type {?} */ (jsonObj["Type"]));
        /** @type {?} */
        let content = jsonObj["Content"];
        if (itemType == CommandItemType.MethodRefer) {
            return Object.assign(new CmpMethodRefering(), (/** @type {?} */ (content)));
            ;
        }
        else if (itemType == CommandItemType.Branch) {
            /** @type {?} */
            let branchConvertor = new BranchCommandItemConvertor();
            return branchConvertor.InitFromJobject(jsonObj["Content"]);
        }
        else if (itemType == CommandItemType.BranchCollection) {
            /** @type {?} */
            let branchCollectionConvertor = new BranchCollectionCommandItemConvertor();
            return branchCollectionConvertor.InitFromJobject(jsonObj["Content"]);
        }
        return null;
    }
}
