// @gsp-cmp/webcommand/esm2015/lib/vm/WebCommandMetadata.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { WebCommandConvertor } from "./Command";
import { ExtendsConvert } from "./ExtendProperty";
export class WebCommandMetadata {
}
if (false) {
    /** @type {?} */
    WebCommandMetadata.prototype.Id;
    /** @type {?} */
    WebCommandMetadata.prototype.Code;
    /** @type {?} */
    WebCommandMetadata.prototype.Name;
    /** @type {?} */
    WebCommandMetadata.prototype.Description;
    /** @type {?} */
    WebCommandMetadata.prototype.Extends;
    /** @type {?} */
    WebCommandMetadata.prototype.Commands;
}
export class WebCommandMetadataConvertor {
    /**
     * @param {?} obj
     * @return {?}
     */
    ConvertJObject(obj) {
        /** @type {?} */
        let metadata = (/** @type {?} */ (obj));
        /** @type {?} */
        let jobj = new Object();
        jobj["Id"] = metadata.Id;
        jobj["Code"] = metadata.Code;
        jobj["Name"] = metadata.Name;
        jobj["Description"] = metadata.Description;
        if (metadata.Commands != null) {
            /** @type {?} */
            let CommandsJArry = jobj["Commands"] = [];
            /** @type {?} */
            let convertor = new WebCommandConvertor();
            metadata.Commands.forEach((/**
             * @param {?} command
             * @return {?}
             */
            command => {
                CommandsJArry.push(convertor.ConvertJObject(command));
            }));
        }
        if (metadata.Extends != null) {
            /** @type {?} */
            let convertor = new ExtendsConvert();
            jobj["Extends"] = convertor.ConvertJObject(metadata.Extends);
        }
        return jobj;
    }
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    InitFromJobject(jsonObj) {
        /** @type {?} */
        let metadata = new WebCommandMetadata();
        metadata.Id = jsonObj["Id"];
        metadata.Code = jsonObj["Code"];
        metadata.Name = jsonObj["Name"];
        metadata.Description = jsonObj["Description"];
        /** @type {?} */
        let CommandsJArry = jsonObj["Commands"];
        if (CommandsJArry != null) {
            metadata.Commands = new Array();
            CommandsJArry.forEach((/**
             * @param {?} element
             * @return {?}
             */
            element => {
                /** @type {?} */
                let cmpOpSerializer = new WebCommandConvertor();
                metadata.Commands.push((/** @type {?} */ (cmpOpSerializer.InitFromJobject(element))));
            }));
        }
        /** @type {?} */
        let convertor = new ExtendsConvert();
        metadata.Extends = convertor.InitFromJobject(jsonObj["Extends"]);
        return metadata;
    }
}
