// @farris/ui-tree/esm5/lib/show-title.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, Injector } from '@angular/core';
var ShowTitleDirective = /** @class */ (function () {
    function ShowTitleDirective(el, injector) {
        this.el = el;
        this.injector = injector;
    }
    /**
     * @return {?}
     */
    ShowTitleDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    ShowTitleDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.enableTips) {
            this.el.nativeElement.title = this.el.nativeElement.innerText;
        }
    };
    ShowTitleDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[show-title]'
                },] }
    ];
    /** @nocollapse */
    ShowTitleDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Injector }
    ]; };
    ShowTitleDirective.propDecorators = {
        enableTips: [{ type: Input, args: ['show-title',] }]
    };
    return ShowTitleDirective;
}());
export { ShowTitleDirective };
if (false) {
    /** @type {?} */
    ShowTitleDirective.prototype.enableTips;
    /**
     * @type {?}
     * @private
     */
    ShowTitleDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    ShowTitleDirective.prototype.injector;
}
