// @edp-aif/runtime-api/esm5/lib/aif-manual-creation/aif-runtime.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, SystemJsNgModuleLoader, NgModuleFactoryLoader, LOCALE_ID } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { ModalModule } from '@farris/ui-modal';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { ReferSelectorComponent } from './components/refer-selector/refer-selector.component';
import { PushSelectorComponent } from './components/push-selector/push-selector.component';
import { AifRuntimeModuleConfig } from './config/aif-runtime-module.config';
import { SourceDataSelectorComponent } from './components/source-data-selector/source-data-selector.component';
import { TreeTableModule } from '@farris/ui-treetable';
import { NotifyModule } from '@farris/ui-notify';
import { MessagerModule } from '@farris/ui-messager';
import { AifLocalePipe } from './pipe/locale.pipe';
var ɵ0 = localStorage.getItem('languageCode') || 'zh-CHS', ɵ1 = {};
/**
 * @deprecated please use \@edp-bif/runtime-api
 */
var AifRuntimeModule = /** @class */ (function () {
    function AifRuntimeModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    AifRuntimeModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            "ngModule": AifRuntimeModule,
            "providers": [
                { provide: AifRuntimeModuleConfig, useValue: config || {} },
            ]
        };
    };
    AifRuntimeModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        HttpClientModule,
                        ModalModule.forRoot(),
                        NotifyModule.forRoot(),
                        MessagerModule.forRoot(),
                        FarrisDialogModule.forRoot(),
                        TreeTableModule,
                    ],
                    providers: [
                        { provide: LOCALE_ID, useValue: ɵ0 },
                        { provide: AifRuntimeModuleConfig, useValue: ɵ1 },
                        { provide: NgModuleFactoryLoader, useClass: SystemJsNgModuleLoader },
                    ],
                    declarations: [
                        ReferSelectorComponent,
                        PushSelectorComponent,
                        SourceDataSelectorComponent,
                        AifLocalePipe,
                    ],
                    exports: [
                        ReferSelectorComponent,
                        PushSelectorComponent,
                        SourceDataSelectorComponent,
                    ],
                    entryComponents: [
                        ReferSelectorComponent,
                        PushSelectorComponent,
                        SourceDataSelectorComponent,
                    ]
                },] }
    ];
    return AifRuntimeModule;
}());
export { AifRuntimeModule };
export { ɵ0, ɵ1 };
