// @gsp-cmp/ubcc-param-data-ui/esm2015/lib/ubcc-param-dimension-ui/ubcc-param-dimension-ui.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ubcc-param-dimension-ui/ubcc-param-dimension-ui.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
import { DataTypeExtensionConst } from '@gsp-cmp/ubcc-param-vo';
import { DataTypeConst } from '@ecp-caf/common-structure';
export class UbccParamDimensionUiComponent {
    constructor() {
        this.dimensionItemDatas = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} dimensionItem
     * @return {?}
     */
    getControlType(dimensionItem) {
        if (!dimensionItem || !dimensionItem.dataType || !dimensionItem.dataType.getKind()) {
            return;
        }
        /** @type {?} */
        let kind = dimensionItem.dataType.getKind();
        switch (kind) {
            case DataTypeConst.DEFAULT_PRIMITIVE:
                return 'input';
            case DataTypeConst.DEFAULT_ENUMERATION:
                return 'dropdownlist';
            case DataTypeExtensionConst.HELP_TYPE:
                return 'lookup';
            default:
                return 'input';
        }
    }
}
UbccParamDimensionUiComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-ubcc-param-dimension-ui',
                template: "<farris-section [mainTitle]=\"'\u7EF4\u5EA6\u4FE1\u606F'\" [enableMaximize]=\"false\">\n  <div *ngFor=\"let item of dimensionItemDatas;let i=index;\" style=\"margin-top: 10px;\">\n    <div class=\"col-12\" [ngSwitch]=\"getControlType(dimensionItemDatas[i])\">\n      <lib-dimension-dropdownlist-control *ngSwitchCase=\"'dropdownlist'\" [(dimensionItem)]=\"dimensionItemDatas[i]\"></lib-dimension-dropdownlist-control>\n      <lib-dimension-lookup-control *ngSwitchCase=\"'lookup'\" [(dimensionItem)]=\"dimensionItemDatas[i]\"></lib-dimension-lookup-control>\n      <lib-dimension-input-control *ngSwitchDefault [(dimensionItem)]=\"dimensionItemDatas[i]\"></lib-dimension-input-control>\n    </div>\n  </div>\n</farris-section>",
                encapsulation: ViewEncapsulation.None,
                styles: [":host{display:flex;flex-direction:column;flex:1}.f-section-header,lib-ubcc-param-dimension-ui .f-section-extend{margin:0 0 .25rem}"]
            }] }
];
/** @nocollapse */
UbccParamDimensionUiComponent.ctorParameters = () => [];
UbccParamDimensionUiComponent.propDecorators = {
    dimensionItemDatas: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    UbccParamDimensionUiComponent.prototype.dimensionItemDatas;
}
