// @farris/extend-file-upload/esm2015/lib/components/upload-progress/upload-progress.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/upload-progress/upload-progress.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
// const statusColorMap = new Map([
//   ['normal', '#388fff'],
//   ['exception', '#f24645'],
//   ['success', '#51bd78']
// ]);
export class UploadProgressComponent {
    constructor() {
        //是否显示进度条信息
        this.showInfo = true;
        this.percent = 0;
        this.maxWidth = 0;
        //存储的状态
        this.cachedStatus = 'normal';
        //计算的状态
        this.inferredStatus = 'normal';
    }
    //获得进度条状态
    /**
     * @return {?}
     */
    get progressStatus() {
        return this.status || this.inferredStatus;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        const { status, percent, successPercent } = changes;
        //状态变化
        if (status) {
            this.cachedStatus = this.status || this.cachedStatus;
        }
        //进度 已完成的分段百分比 变化
        if (percent || successPercent) {
            /** @type {?} */
            const fillAll = parseInt(this.percent.toString(), 10) >= 100;
            if (fillAll) {
                if ((this.isNotNil(this.successPercent) && (/** @type {?} */ (this.successPercent)) >= 100) || this.successPercent === undefined) {
                    this.inferredStatus = 'success';
                    // this.status = this.inferredStatus;
                }
            }
            else {
                this.inferredStatus = this.cachedStatus;
            }
            this.status = this.inferredStatus;
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @template T
     * @param {?} value
     * @return {?}
     */
    isNotNil(value) {
        return typeof value !== 'undefined' && value !== null;
    }
}
UploadProgressComponent.decorators = [
    { type: Component, args: [{
                selector: 'upload-progress',
                template: "<div [ngClass]=\"'f-progress upload-progress upload-progress-status-' + progressStatus+' f-progress-status-'+progressStatus\" class=\"upload-progress-line\"\r\n  [class.upload-progress-show-info]=\"showInfo\">\r\n  <div class=\"upload-progress-outer\" [ngStyle]=\"{'max-width':maxWidth>0?maxWidth+'px':'auto'}\">\r\n    <div class=\"upload-progress-inner\">\r\n      <div class=\"upload-progress-bg f-progress-bg\" [style.width.%]=\"percent\">\r\n      </div>\r\n      <div *ngIf=\"successPercent || successPercent === 0\" class=\"upload-progress-success-bg f-progress-success-bg\"\r\n        [style.width.%]=\"successPercent\"></div>\r\n    </div>\r\n  </div>\r\n  <span class=\"upload-progress-text\" *ngIf=\"showInfo\">\r\n    <ng-container *ngIf=\"status === 'error' || status === 'success'; else formatTemplate\">\r\n      <span [ngClass]=\"'upload-progress-text-icon f-icon f-icon-'+ status\"></span>\r\n    </ng-container>\r\n  </span>\r\n</div>\r\n<ng-template #formatTemplate>\r\n  {{ percent }}%\r\n</ng-template>",
                styles: [".upload-progress{display:inline-block;font-size:.875rem;color:rgba(0,0,0,.65)}.upload-progress.upload-progress-line{position:relative;width:100%}.upload-progress.upload-progress-line .upload-progress-text .upload-progress-text-icon{font-size:1.25rem}.upload-progress .upload-progress-outer{display:inline-block;width:100%}.upload-progress .upload-progress-inner{position:relative;display:inline-block;width:100%;overflow:hidden;vertical-align:middle;background-color:#efefef;border-radius:100px}.upload-progress .upload-progress-bg,.upload-progress .upload-progress-success-bg{position:relative;border-radius:100px;height:6px;-webkit-transition:.4s cubic-bezier(.08,.82,.17,1);transition:.4s cubic-bezier(.08,.82,.17,1)}.upload-progress .upload-progress-success-bg{position:absolute;top:0;left:0}.upload-progress .upload-progress-text{display:inline-block;width:1.75rem;margin-left:.5rem;color:rgba(0,0,0,.65);font-size:1em;line-height:1;white-space:nowrap;text-align:left;vertical-align:middle;word-break:normal}.upload-progress.upload-progress-status-active .upload-progress-bg::before{position:absolute;top:0;right:0;bottom:0;left:0;background:#fff;border-radius:.625rem;opacity:0;-webkit-animation:2.4s ease-in-out infinite upload-progress-active;animation:2.4s ease-in-out infinite upload-progress-active;content:\"\"}.upload-progress-show-info .upload-progress-outer{padding-right:48px;margin-right:-48px}@-webkit-keyframes upload-progress-active{0%{width:0;opacity:.1}20%{width:0;opacity:.5}100%{width:100%;opacity:0}}@keyframes upload-progress-active{0%{width:0;opacity:.1}20%{width:0;opacity:.5}100%{width:100%;opacity:0}}"]
            }] }
];
/** @nocollapse */
UploadProgressComponent.ctorParameters = () => [];
UploadProgressComponent.propDecorators = {
    showInfo: [{ type: Input }],
    status: [{ type: Input }],
    successPercent: [{ type: Input }],
    percent: [{ type: Input }],
    maxWidth: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    UploadProgressComponent.prototype.showInfo;
    /** @type {?} */
    UploadProgressComponent.prototype.status;
    /** @type {?} */
    UploadProgressComponent.prototype.successPercent;
    /** @type {?} */
    UploadProgressComponent.prototype.percent;
    /** @type {?} */
    UploadProgressComponent.prototype.maxWidth;
    /**
     * @type {?}
     * @private
     */
    UploadProgressComponent.prototype.cachedStatus;
    /**
     * @type {?}
     * @private
     */
    UploadProgressComponent.prototype.inferredStatus;
}
