// @farris/ui-datatable/esm2015/lib/table/cell-style.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { DataTableBodyComponent } from './datatable-body.component';
export class CellStylerDirective {
    /**
     * @param {?} dtb
     * @param {?} elRef
     * @param {?} render
     */
    constructor(dtb, elRef, render) {
        this.dtb = dtb;
        this.elRef = elRef;
        this.render = render;
        this.options = null;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.dtb && this.options) {
            const { val, row, col, index } = this.options;
            /** @type {?} */
            let _sty = null;
            if (!col) {
                _sty = this.dtb.renderRowStyle(row, index) || {};
            }
            else {
                /** @type {?} */
                let rowStyle = this.dtb.renderRowStyle(row, index) || {};
                /** @type {?} */
                let cellStyle = this.dtb.renderCellStyle(val, row, index, col) || {};
                _sty = cellStyle;
                if (col.fixed === 'right') {
                    /** @type {?} */
                    const hasBackground = Object.keys(cellStyle).find((/**
                     * @param {?} n
                     * @return {?}
                     */
                    n => {
                        return ['background', 'background-color', 'backgroundcolor'].indexOf(n.toLowerCase()) > -1;
                    }));
                    if (hasBackground !== null) {
                        _sty = Object.assign(cellStyle, rowStyle);
                    }
                }
            }
            if (_sty) {
                Object.keys(_sty).forEach((/**
                 * @param {?} n
                 * @return {?}
                 */
                n => {
                    this.elRef.nativeElement.style.setProperty(n, _sty[n], 'important');
                }));
            }
        }
    }
}
CellStylerDirective.decorators = [
    { type: Directive, args: [{ selector: '[cell-styler]' },] }
];
/** @nocollapse */
CellStylerDirective.ctorParameters = () => [
    { type: DataTableBodyComponent },
    { type: ElementRef },
    { type: Renderer2 }
];
CellStylerDirective.propDecorators = {
    options: [{ type: Input, args: ['cell-styler',] }]
};
if (false) {
    /** @type {?} */
    CellStylerDirective.prototype.options;
    /**
     * @type {?}
     * @private
     */
    CellStylerDirective.prototype.dtb;
    /**
     * @type {?}
     * @private
     */
    CellStylerDirective.prototype.elRef;
    /**
     * @type {?}
     * @private
     */
    CellStylerDirective.prototype.render;
}
