// @farris/ui-datatable/esm2015/lib/table/datatable-footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ContentChild, TemplateRef } from '@angular/core';
export class DatatableFooterComponent {
    constructor() {
    }
}
DatatableFooterComponent.decorators = [
    { type: Component, args: [{
                selector: 'datatable-footer',
                template: `
    <div class="table"
    [class.table-hover]="hover">
        <ng-container *ngIf="!tableFooter">
                <table class="table table-hover">
                        <colgroup>
                            <col class="dt-checkbox-cell" />
                            <col *ngFor="let col of columns" [style.width]="col.width + 'px'" />
                        </colgroup>
                        <thead>
                            <tr>
                                <th class="dt-checkbox-cell">
                                    
                                </th>
                                <th  *ngFor="let col of columns" [attr.align]="col.align" >{{ col.title }}</th>
                            </tr>
                        </thead>
                    </table>
        </ng-container>
</div>
    `
            }] }
];
/** @nocollapse */
DatatableFooterComponent.ctorParameters = () => [];
DatatableFooterComponent.propDecorators = {
    hover: [{ type: Input }],
    columns: [{ type: Input }],
    tableFooter: [{ type: Input }, { type: ContentChild, args: [TemplateRef,] }]
};
if (false) {
    /** @type {?} */
    DatatableFooterComponent.prototype.hover;
    /** @type {?} */
    DatatableFooterComponent.prototype.columns;
    /** @type {?} */
    DatatableFooterComponent.prototype.tableFooter;
}
