// @gsp-svc/file-viewer/esm2015/lib/gsp-document/gsp-page/gsp-page.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
export class GspPageComponent {
    constructor() {
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        /** @type {?} */
        const isIE = /*@cc_on!@*/ false || !!/(MSIE|Trident\/|Edge\/)/i.test(navigator.userAgent);
        if (isIE && this.number === 0) {
            this.editable = false;
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (this.isHtml) {
            // TODO: this is temporary needed to remove unneeded spaces and BOM symbol 
            // which leads to undesired spaces on the top of the docs pages
            this.data = this.data
                ? this.data.replace(/>\s+</g, '><')
                    .replace(/\uFEFF/g, "")
                : null;
        }
        else {
            if (this.data) {
                this.imgData = this.data.startsWith('data:image') ? this.data : 'data:image/png;base64,' + this.data;
            }
        }
    }
}
GspPageComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-gsp-page',
                template: "<div id=\"page-{{number}}\">\n  <div class=\"gsp-wrapper\" [innerHTML]=\"data | safeHtml\" *ngIf=\"data && isHtml\" [contentEditable]=\"(editable) ? true : false\"></div>\n  <img class=\"gsp-page-image\" [style.width.px]=\"width\" [style.height.px]=\"height\" [attr.src]=\"imgData | safeResourceHtml\"\n       alt=\"\"\n       *ngIf=\"data && !isHtml\">\n  <div class=\"gsp-page-spinner\" *ngIf=\"!data\">\n  </div>\n</div>",
                styles: [".gsp-page-spinner{margin-top:150px;text-align:center}.gsp-wrapper{width:inherit;height:inherit}.gsp-wrapper img{width:inherit}.gsp-wrapper div{width:100%}.gsp-highlight{background-color:#ff0}.gsp-highlight-select{background-color:#ff9b00}.gsp-page-image{height:100%!important;width:100%!important}"]
            }] }
];
/** @nocollapse */
GspPageComponent.ctorParameters = () => [];
GspPageComponent.propDecorators = {
    angle: [{ type: Input }],
    width: [{ type: Input }],
    height: [{ type: Input }],
    number: [{ type: Input }],
    data: [{ type: Input }],
    isHtml: [{ type: Input }],
    editable: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    GspPageComponent.prototype.angle;
    /** @type {?} */
    GspPageComponent.prototype.width;
    /** @type {?} */
    GspPageComponent.prototype.height;
    /** @type {?} */
    GspPageComponent.prototype.number;
    /** @type {?} */
    GspPageComponent.prototype.data;
    /** @type {?} */
    GspPageComponent.prototype.isHtml;
    /** @type {?} */
    GspPageComponent.prototype.editable;
    /** @type {?} */
    GspPageComponent.prototype.imgData;
}
