// @farris/dynamic-control-group/esm2015/lib/components/dropdown/base-dropdown.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from "@angular/core";
import { FormGroup } from "@angular/forms";
export class BaseDropdownComponent {
    constructor() { }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.options = this.control.control;
        this.className = this.control.className ? this.control.className : 'col-12 col-md-6 col-xl-3 col-el-2';
    }
}
BaseDropdownComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-base-dropdown",
                template: `
    <div class="farris-group-wrap" [formGroup]="form">
      <div class="form-group farris-form-group">
        <label class="col-form-label">
          <span *ngIf="0" class="farris-label-info text-danger">*</span>
          <span class="farris-label-text">{{control.name}}</span>
        </label>
        <div class="farris-input-wrap">
            <farris-combo-list *ngIf="options.items"
                [readonly]="options.readonly"
                [formControlName]="control.labelCode"
                [idField]="options.idField"
                [textField]="options.textField"
                [multiSelect]="options.multiSelect"
                [data]="options.items"
            ></farris-combo-list>
            <farris-combo-list *ngIf="options.uri"
                [readonly]="options.readonly"
                [formControlName]="control.labelCode"
                [idField]="options.idField"
                [textField]="options.textField"
                [multiSelect]="options.multiSelect"
                [uri]="options.uri"
            ></farris-combo-list>
        </div>
      </div>
    </div>
  `
            }] }
];
/** @nocollapse */
BaseDropdownComponent.ctorParameters = () => [];
BaseDropdownComponent.propDecorators = {
    className: [{ type: HostBinding, args: ['class',] }],
    control: [{ type: Input }],
    form: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    BaseDropdownComponent.prototype.className;
    /** @type {?} */
    BaseDropdownComponent.prototype.control;
    /** @type {?} */
    BaseDropdownComponent.prototype.form;
    /** @type {?} */
    BaseDropdownComponent.prototype.options;
}
