// @farris/ui-search-box/esm2015/lib/search-fields/search-filed.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ApplicationRef, ComponentFactoryResolver, Directive, ElementRef, HostListener, Injector, Input, NgZone, Renderer2 } from '@angular/core';
import { debounceTime } from 'rxjs/operators';
import { SearchFieldPanelComponent } from './search-field-panel.component';
import { SearchFieldsComponent } from './search-fields.component';
import { CommonUtils } from '@farris/ui-common';
export class SearchFieldDirective {
    /**
     * @param {?} injector
     * @param {?} elRef
     * @param {?} ngzone
     * @param {?} render
     * @param {?} cfr
     * @param {?} searchFieldsRef
     * @param {?} _applicationRef
     */
    constructor(injector, elRef, ngzone, render, cfr, searchFieldsRef, _applicationRef) {
        this.injector = injector;
        this.elRef = elRef;
        this.ngzone = ngzone;
        this.render = render;
        this.cfr = cfr;
        this.searchFieldsRef = searchFieldsRef;
        this._applicationRef = _applicationRef;
        this.winResizeHandle = null;
        this.docKeydownEventHandle = null;
        this.commonUtils = null;
        this.commonUtils = this.injector.get(CommonUtils, new CommonUtils());
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.searchFieldsRef) {
            this.searchFieldsRef.hideShadowbox.subscribe((/**
             * @param {?} e
             * @return {?}
             */
            (e) => {
                this.removeFieldPanel();
            }));
        }
        this.winResizeHandle = this.render.listen(window, 'resize', (/**
         * @return {?}
         */
        () => {
            this.removeFieldPanel();
        }));
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.removeFieldPanel();
        if (this.winResizeHandle) {
            this.winResizeHandle();
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onMouseDown($event) {
        $event.stopPropagation();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onFieldClick($event) {
        $event.stopPropagation();
        if (!this.panelElement) {
            this.createFieldPanel();
        }
        else {
            this.removeFieldPanel();
        }
    }
    /**
     * @private
     * @return {?}
     */
    removeFieldPanel() {
        if (this.panelElement) {
            this.panelElement.remove();
            this.panelElement = null;
            this.searchFieldsRef.overLayService.destory(this.elRef.nativeElement);
            this.render.removeClass(this.elRef.nativeElement, 'selected');
        }
        if (this.docKeydownEventHandle) {
            this.docKeydownEventHandle();
            this.docKeydownEventHandle = null;
        }
    }
    /**
     * @private
     * @param {?=} enterFn
     * @return {?}
     */
    registerKeyboardEvent(enterFn) {
        if (this.commonUtils) {
            return this.commonUtils.regBodyKeydownEvent(enterFn, (/**
             * @return {?}
             */
            () => { this.removeFieldPanel(); }));
        }
        return null;
    }
    /**
     * @private
     * @return {?}
     */
    createFieldPanel() {
        this.panelElement = document.createElement('div');
        this.panelElement.classList.add('overlay-pane', 'f-filter-panel', 'f-area-hide', 'f-search-field-container');
        /** @type {?} */
        const zindex = '' + this.searchFieldsRef.commonUtils.getFloatingLayerIndex();
        this.panelElement.style.zIndex = zindex;
        document.body.appendChild(this.panelElement);
        this.setPanelPosition();
        /** @type {?} */
        const fieldPanelRef = this.cfr.resolveComponentFactory(SearchFieldPanelComponent);
        /** @type {?} */
        const _fieldPanelIns = fieldPanelRef.create(this.injector);
        this._applicationRef.attachView(_fieldPanelIns.hostView);
        _fieldPanelIns.instance.fieldOptions = this.field;
        _fieldPanelIns.instance.conditions = { field: this.field.code, value: this.field.value };
        this.panelElement.appendChild(_fieldPanelIns.location.nativeElement);
        this.panelElement.classList.add('f-area-show');
        _fieldPanelIns.instance.filterChange.pipe(debounceTime(100)).subscribe((/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            // 新的查询条件
            // console.log(e);
            this.searchFieldsRef.setConditions(e);
        }));
        _fieldPanelIns.instance.remove.subscribe((/**
         * @param {?} e
         * @return {?}
         */
        e => {
            this.searchFieldsRef.remove(e.code);
        }));
        _fieldPanelIns.changeDetectorRef.detectChanges();
        this.searchFieldsRef.overLayService.registerMouseEvent(this.elRef.nativeElement, (/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            if (this.elRef.nativeElement.contains(e.target)) {
                return;
            }
            if (e.target.closest('.f-search-field-container')) {
                if (e.target.nodeName === 'BUTTON' && e.target.className.indexOf('close-field-panel') > -1) {
                    this.removeFieldPanel();
                }
                return;
            }
            if (e.target.closest('.date-overlay-container') || e.target.className.indexOf('date-overlay-container') > -1) {
                return;
            }
            this.removeFieldPanel();
        }));
        this.render.addClass(this.elRef.nativeElement, 'selected');
        // 处理 ESC、ENTER
        this.docKeydownEventHandle = this.registerKeyboardEvent((/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const btnok = _fieldPanelIns.instance.btnSubmit;
            if (btnok && !btnok.nativeElement.disabled) {
                btnok.nativeElement.click();
                this.removeFieldPanel();
            }
        }));
    }
    /**
     * @private
     * @param {?=} updateTopPosition
     * @return {?}
     */
    setPanelPosition(updateTopPosition = true) {
        if (this.panelElement) {
            const { width, left, top, height } = this.getPanelSize();
            this.panelElement.style.width = `${width}px`;
            // this.panelElement.style.maxHeight = `${height}px`;
            if (updateTopPosition) {
                this.panelElement.style.top = `${top}px`;
            }
            this.panelElement.style.left = `${left}px`;
        }
    }
    /**
     * @private
     * @return {?}
     */
    getInputSizeInfo() {
        return this.elRef.nativeElement.getBoundingClientRect();
    }
    /**
     * @private
     * @return {?}
     */
    getPanelSize() {
        /** @type {?} */
        const width = 380;
        let { top, left, height } = this.getInputSizeInfo();
        /** @type {?} */
        const bottom = window.innerHeight - height - top;
        /** @type {?} */
        let panelHeight = 190;
        /** @type {?} */
        const h = top > bottom ? top : bottom;
        if (bottom > panelHeight) {
            top = top + height;
            // 面板由上向下展开
            this.panelElement.style.transformOrigin = '100% top';
        }
        else {
            if (top > bottom) {
                if (h < panelHeight) {
                    panelHeight = h - 10;
                    top = 10;
                }
                else {
                    top = top - parseInt('' + panelHeight, 10) - 5;
                }
                // 面板由下向上展开
                this.panelElement.style.transformOrigin = '100% bottom';
            }
            else {
                if (h < panelHeight) {
                    panelHeight = h - 10;
                }
                top = top + height;
            }
        }
        if (window.innerWidth - left < width) {
            left = left + width - width;
        }
        return { width, top, height: panelHeight, left };
    }
}
SearchFieldDirective.decorators = [
    { type: Directive, args: [{ selector: '[search-field]' },] }
];
/** @nocollapse */
SearchFieldDirective.ctorParameters = () => [
    { type: Injector },
    { type: ElementRef },
    { type: NgZone },
    { type: Renderer2 },
    { type: ComponentFactoryResolver },
    { type: SearchFieldsComponent },
    { type: ApplicationRef }
];
SearchFieldDirective.propDecorators = {
    field: [{ type: Input, args: ['search-field',] }],
    onMouseDown: [{ type: HostListener, args: ['mousedown', ['$event'],] }],
    onFieldClick: [{ type: HostListener, args: ['click', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    SearchFieldDirective.prototype.field;
    /**
     * @type {?}
     * @private
     */
    SearchFieldDirective.prototype.panelElement;
    /**
     * @type {?}
     * @private
     */
    SearchFieldDirective.prototype.winResizeHandle;
    /**
     * @type {?}
     * @private
     */
    SearchFieldDirective.prototype.docKeydownEventHandle;
    /**
     * @type {?}
     * @private
     */
    SearchFieldDirective.prototype.commonUtils;
    /**
     * @type {?}
     * @private
     */
    SearchFieldDirective.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    SearchFieldDirective.prototype.elRef;
    /**
     * @type {?}
     * @private
     */
    SearchFieldDirective.prototype.ngzone;
    /**
     * @type {?}
     * @private
     */
    SearchFieldDirective.prototype.render;
    /**
     * @type {?}
     * @private
     */
    SearchFieldDirective.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    SearchFieldDirective.prototype.searchFieldsRef;
    /**
     * @type {?}
     * @private
     */
    SearchFieldDirective.prototype._applicationRef;
}
