// @farris/dynamic-control-group/esm5/lib/components/dropdown/base-dropdown.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from "@angular/core";
import { FormGroup } from "@angular/forms";
var BaseDropdownComponent = /** @class */ (function () {
    function BaseDropdownComponent() {
    }
    /**
     * @return {?}
     */
    BaseDropdownComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.options = this.control.control;
        this.className = this.control.className ? this.control.className : 'col-12 col-md-6 col-xl-3 col-el-2';
    };
    BaseDropdownComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-base-dropdown",
                    template: "\n    <div class=\"farris-group-wrap\" [formGroup]=\"form\">\n      <div class=\"form-group farris-form-group\">\n        <label class=\"col-form-label\">\n          <span *ngIf=\"0\" class=\"farris-label-info text-danger\">*</span>\n          <span class=\"farris-label-text\">{{control.name}}</span>\n        </label>\n        <div class=\"farris-input-wrap\">\n            <farris-combo-list *ngIf=\"options.items\"\n                [readonly]=\"options.readonly\"\n                [formControlName]=\"control.labelCode\"\n                [idField]=\"options.idField\"\n                [textField]=\"options.textField\"\n                [multiSelect]=\"options.multiSelect\"\n                [data]=\"options.items\"\n            ></farris-combo-list>\n            <farris-combo-list *ngIf=\"options.uri\"\n                [readonly]=\"options.readonly\"\n                [formControlName]=\"control.labelCode\"\n                [idField]=\"options.idField\"\n                [textField]=\"options.textField\"\n                [multiSelect]=\"options.multiSelect\"\n                [uri]=\"options.uri\"\n            ></farris-combo-list>\n        </div>\n      </div>\n    </div>\n  "
                }] }
    ];
    /** @nocollapse */
    BaseDropdownComponent.ctorParameters = function () { return []; };
    BaseDropdownComponent.propDecorators = {
        className: [{ type: HostBinding, args: ['class',] }],
        control: [{ type: Input }],
        form: [{ type: Input }]
    };
    return BaseDropdownComponent;
}());
export { BaseDropdownComponent };
if (false) {
    /** @type {?} */
    BaseDropdownComponent.prototype.className;
    /** @type {?} */
    BaseDropdownComponent.prototype.control;
    /** @type {?} */
    BaseDropdownComponent.prototype.form;
    /** @type {?} */
    BaseDropdownComponent.prototype.options;
}
