// @gsp-sys/sysmgr-common/esm5/lib/util/gsp-sec-setting.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 保密测评相关配置
 */
var /**
 * 保密测评相关配置
 */
GspSecSetting = /** @class */ (function () {
    function GspSecSetting() {
        this.id = '';
        this.funcObj = '';
        this.userType = '';
        this.funcOp = '';
        this.funcPage = '';
        this.canMgrUserTypeScope = '';
    }
    return GspSecSetting;
}());
/**
 * 保密测评相关配置
 */
export { GspSecSetting };
if (false) {
    /** @type {?} */
    GspSecSetting.prototype.id;
    /**
     * 功能菜单标识
     * @type {?}
     */
    GspSecSetting.prototype.funcObj;
    /**
     * 用户角色
     * @type {?}
     */
    GspSecSetting.prototype.userType;
    /**
     * 操作类型，view，edit等
     * @type {?}
     */
    GspSecSetting.prototype.funcOp;
    /**
     * 功能配置页签
     * @type {?}
     */
    GspSecSetting.prototype.funcPage;
    /**
     * 可以管理的用户范围
     * @type {?}
     */
    GspSecSetting.prototype.canMgrUserTypeScope;
}
