// @gsp-sys/sysmgr-common/esm5/lib/util/result-package.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 二元关系实体增量包
 */
var /**
 * 二元关系实体增量包
 */
ResultPackage = /** @class */ (function () {
    function ResultPackage() {
        /**
         * 是否进行校验
         */
        this.check = true;
        this.removingResults = new Array();
        this.addingResults = new Array();
        this.ownerAuditEntity = { id: null, code: null, name: null };
        this.addingEntity = new Array();
        this.removingEntity = new Array();
        this.allFlag = '0';
    }
    return ResultPackage;
}());
/**
 * 二元关系实体增量包
 */
export { ResultPackage };
if (false) {
    /**
     * 主体内码
     * @type {?}
     */
    ResultPackage.prototype.ownerId;
    /**
     * 授权的客体删除列表
     * @type {?}
     */
    ResultPackage.prototype.removingResults;
    /**
     * 授权的客体增加列表
     * @type {?}
     */
    ResultPackage.prototype.addingResults;
    /**
     * 主体描述
     * @type {?}
     */
    ResultPackage.prototype.ownerAuditEntity;
    /**
     * 新增客体描述
     * @type {?}
     */
    ResultPackage.prototype.addingEntity;
    /**
     * 删除客体描述
     * @type {?}
     */
    ResultPackage.prototype.removingEntity;
    /**
     * 是否授权所有，0=否，1=所有
     * @type {?}
     */
    ResultPackage.prototype.allFlag;
    /**
     * 是否进行校验
     * @type {?}
     */
    ResultPackage.prototype.check;
}
