// @gsp-lcm/metadatart-selector/esm5/metadata/metadatart-select/metadatart-select.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadatart-select/metadatart-select.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SelectMetadataRTComponent } from './components/selectmetadatart/selectmetadatart.component';
import { TemplatemetadataRTComponent } from './components/metadatartlist/metadatartlist.component';
import { FormsModule } from '@angular/forms';
import { TreeViewModule } from '@progress/kendo-angular-treeview';
import { GridModule } from '@progress/kendo-angular-grid';
import { MetadataRTService } from './metadatart.service';
import { NotifyModule } from '@farris/ui-notify';
// import { FarrisDialogModule } from '@farris/ui-dialog';
var MetadataRTSelectModule = /** @class */ (function () {
    function MetadataRTSelectModule() {
    }
    MetadataRTSelectModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        TreeViewModule,
                        GridModule,
                        NotifyModule.forRoot()
                        // FarrisDialogModule
                    ],
                    exports: [
                        SelectMetadataRTComponent,
                        NotifyModule
                    ],
                    providers: [
                        MetadataRTService
                    ],
                    // 服务提供者
                    declarations: [SelectMetadataRTComponent, TemplatemetadataRTComponent],
                    entryComponents: [SelectMetadataRTComponent]
                },] }
    ];
    /** @nocollapse */
    MetadataRTSelectModule.ctorParameters = function () { return []; };
    return MetadataRTSelectModule;
}());
export { MetadataRTSelectModule };
