// @gsp-lcm/metadatart-selector/esm5/metadata/metadatart-select/metadatart.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadatart-select/metadatart.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { GSPMetadataRTService } from '../metadata.rtservice';
var MetadataRTService = /** @class */ (function () {
    function MetadataRTService(gspMetadataService) {
        this.gspMetadataService = gspMetadataService;
        this.metadatas = [];
    }
    /**
     * @param {?} metadata
     * @return {?}
     */
    MetadataRTService.prototype.addMetadataRT = /**
     * @param {?} metadata
     * @return {?}
     */
    function (metadata) {
        this.metadatas.push(metadata);
    };
    /**
     * @return {?}
     */
    MetadataRTService.prototype.clearMetadataRT = /**
     * @return {?}
     */
    function () {
        this.metadatas = [];
    };
    // 根据类型获取Cloud Server下所有的元数据
    // 根据类型获取Cloud Server下所有的元数据
    /**
     * @param {?} typeList
     * @return {?}
     */
    MetadataRTService.prototype.GetMetadataRTList = 
    // 根据类型获取Cloud Server下所有的元数据
    /**
     * @param {?} typeList
     * @return {?}
     */
    function (typeList) {
        return this.gspMetadataService.GetMetadataRTList(typeList);
    };
    /**
     * @param {?} typeList
     * @return {?}
     */
    MetadataRTService.prototype.GetMetadataRTListForHelp = /**
     * @param {?} typeList
     * @return {?}
     */
    function (typeList) {
        return this.gspMetadataService.GetMetadataRTListForHelp(typeList);
    };
    /**
     * @param {?} bizTypeId
     * @param {?} metadataTypes
     * @return {?}
     */
    MetadataRTService.prototype.getMetadataListByBizTypeId = /**
     * @param {?} bizTypeId
     * @param {?} metadataTypes
     * @return {?}
     */
    function (bizTypeId, metadataTypes) {
        return this.gspMetadataService.getMetadataListByBizTypeId(bizTypeId, metadataTypes);
    };
    /**
     * @param {?} boList
     * @param {?} bizTypeId
     * @param {?} metadataTypes
     * @return {?}
     */
    MetadataRTService.prototype.getMetadataListByBizTypeIdForHelp = /**
     * @param {?} boList
     * @param {?} bizTypeId
     * @param {?} metadataTypes
     * @return {?}
     */
    function (boList, bizTypeId, metadataTypes) {
        if (boList === undefined) {
            boList = '';
        }
        if (bizTypeId === undefined) {
            bizTypeId = '';
        }
        return this.gspMetadataService.getMetadataListByBizTypeIdForHelp(boList, bizTypeId, metadataTypes);
    };
    // 获根据Id取指定元数据
    // 获根据Id取指定元数据
    /**
     * @param {?} metadataID
     * @return {?}
     */
    MetadataRTService.prototype.GetMetadataRT = 
    // 获根据Id取指定元数据
    /**
     * @param {?} metadataID
     * @return {?}
     */
    function (metadataID) {
        return this.gspMetadataService.GetMetadataRT(metadataID);
    };
    MetadataRTService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MetadataRTService.ctorParameters = function () { return [
        { type: GSPMetadataRTService }
    ]; };
    return MetadataRTService;
}());
export { MetadataRTService };
if (false) {
    /** @type {?} */
    MetadataRTService.prototype.metadatas;
    /**
     * @type {?}
     * @private
     */
    MetadataRTService.prototype.gspMetadataService;
}
