// @qdp/ide-cmp/esm2015/lib/service/paras-mapping/paras-mapping.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { ParasMappingManagerService } from '@qdp/common';
export class ParasMappingService {
    /**
     * @param {?} parasMappingManager
     */
    constructor(parasMappingManager) {
        this.parasMappingManager = parasMappingManager;
        this.paramMappingInfo = new Subject();
        this.parasMappingManager.parasmappingInfoChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            this.paramMappingInfo.next(value);
        }));
    }
    /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    getParasMapping(id, queryId, queryRelativeUrl) {
        return this.parasMappingManager.getParasMapping(id, queryId, queryRelativeUrl);
    }
    /**
     * @param {?} paramMappingInfo
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    saveParasMapping(paramMappingInfo, queryRelativeUrl) {
        return this.parasMappingManager.saveParasMapping(paramMappingInfo, queryRelativeUrl);
    }
    /**
     * @param {?} datasourceID
     * @param {?} queryID
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    deleteParasMapping(datasourceID, queryID, queryRelativeUrl) {
        return this.parasMappingManager.deleteParasMapping(datasourceID, queryID, queryRelativeUrl);
    }
}
ParasMappingService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ParasMappingService.ctorParameters = () => [
    { type: ParasMappingManagerService }
];
if (false) {
    /** @type {?} */
    ParasMappingService.prototype.paramMappingInfo;
    /**
     * @type {?}
     * @private
     */
    ParasMappingService.prototype.parasMappingManager;
}
