// @gsp-sys/rtf-common/esm5/lib/services/i18nService/i18n-setting.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpService, LanguageService } from '@ecp-caf/caf-common';
import { of } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import { FrameworkVariableService } from '../frmVariableService/framework-variable.service';
/**
 * @record
 */
export function I18nSettingEntity() { }
if (false) {
    /** @type {?|undefined} */
    I18nSettingEntity.prototype.i18nSetting;
    /** @type {?|undefined} */
    I18nSettingEntity.prototype.languages;
    /** @type {?|undefined} */
    I18nSettingEntity.prototype.currentLanCode;
}
var FrmI18nSettingService = /** @class */ (function () {
    function FrmI18nSettingService(frmVariable, lanService, http) {
        this.frmVariable = frmVariable;
        this.lanService = lanService;
        this.http = http;
        this.variableKey = 'frm-i18nSetting';
    }
    /**
     * @return {?}
     */
    FrmI18nSettingService.prototype.getSetting = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var temp = this.frmVariable.getFrmVariableByKey(this.variableKey);
        return this.deepClone(temp);
    };
    /**
     * @return {?}
     */
    FrmI18nSettingService.prototype.getSetting$ = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var temp = this.frmVariable.getFrmVariableByKey(this.variableKey);
        return !temp ? this.loadSetting$() : of(this.deepClone(temp));
    };
    /**
     * @return {?}
     */
    FrmI18nSettingService.prototype.loadSetting$ = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var rtfService = ((/** @type {?} */ (window))).gspframeworkService && ((/** @type {?} */ (window))).gspframeworkService.rtf;
        /** @type {?} */
        var entity = {};
        if (!!rtfService && !!rtfService.language) {
            try {
                /** @type {?} */
                var lanService = rtfService.language;
                lanService.i18nSetting((/**
                 * @param {?} res
                 * @return {?}
                 */
                function (res) {
                    entity.i18nSetting = res;
                }), (/**
                 * @param {?} err
                 * @return {?}
                 */
                function (err) {
                    console.log(err);
                }));
                lanService.allLanguages((/**
                 * @param {?} res
                 * @return {?}
                 */
                function (res) {
                    entity.languages = res;
                }), (/**
                 * @param {?} err
                 * @return {?}
                 */
                function (err) {
                    console.log(err);
                }));
                entity.currentLanCode = lanService.getLanguageCode();
                this.frmVariable.frameworkVariableReg(this.variableKey, entity);
                return of(entity);
            }
            catch (e) {
                console.log(e);
                return of(null);
            }
        }
        else {
            /** @type {?} */
            var languagesUrl = '/api/runtime/sys/v1.0/i18n/languageservice/getalllanguages';
            /** @type {?} */
            var settingUrl_1 = '/api/runtime/sys/v1.0/i18n/i18nusersetting/setting';
            entity.currentLanCode = this.lanService.getLanguageCode();
            return this.http.get(languagesUrl).pipe(map((/**
             * @param {?} res
             * @return {?}
             */
            function (res) { return entity.languages = res; })), switchMap((/**
             * @return {?}
             */
            function () { return _this.http.get(settingUrl_1); })), map((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                entity.i18nSetting = res;
                _this.frmVariable.frameworkVariableReg(_this.variableKey, entity);
                return entity;
            })), catchError((/**
             * @param {?} err
             * @return {?}
             */
            function (err) {
                console.log(err);
                return of(null);
            })));
        }
    };
    /**
     * @private
     * @param {?} target
     * @return {?}
     */
    FrmI18nSettingService.prototype.deepClone = /**
     * @private
     * @param {?} target
     * @return {?}
     */
    function (target) {
        /** @type {?} */
        var result;
        if (typeof target === 'object') {
            if (Array.isArray(target)) {
                result = [];
                for (var i in target) {
                    result.push(this.deepClone(target[i]));
                }
            }
            else if (target === null) {
                result = null;
            }
            else if (target.constructor === RegExp) {
                result = target;
            }
            else {
                result = {};
                for (var i in target) {
                    result[i] = this.deepClone(target[i]);
                }
            }
        }
        else {
            result = target;
        }
        return result;
    };
    FrmI18nSettingService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FrmI18nSettingService.ctorParameters = function () { return [
        { type: FrameworkVariableService },
        { type: LanguageService },
        { type: HttpService }
    ]; };
    return FrmI18nSettingService;
}());
export { FrmI18nSettingService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FrmI18nSettingService.prototype.variableKey;
    /**
     * @type {?}
     * @private
     */
    FrmI18nSettingService.prototype.frmVariable;
    /**
     * @type {?}
     * @private
     */
    FrmI18nSettingService.prototype.lanService;
    /**
     * @type {?}
     * @private
     */
    FrmI18nSettingService.prototype.http;
}
