// @gsp-wf/wf-process-editor/esm5/lib/views/performer/views/matrix-performer/service/matrix-performer.component.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { map } from 'rxjs/operators';
var MatrixDefHelpService = /** @class */ (function () {
    function MatrixDefHelpService(http) {
        this.http = http;
    }
    /**
     * @param {?} url
     * @param {?=} params
     * @return {?}
     */
    MatrixDefHelpService.prototype.getData = /**
     * @param {?} url
     * @param {?=} params
     * @return {?}
     */
    function (url, params) {
        if (url.indexOf('getMatrixDef') > -1) {
            /** @type {?} */
            var index = params.pageIndex || 1;
            /** @type {?} */
            var size = params.pageSize || 20;
            /** @type {?} */
            var searchValue = JSON.parse(params.searchValue);
            if (searchValue.searchValue) {
                /** @type {?} */
                var query = '{';
                if (searchValue.searchValue) {
                    if (searchValue.searchField === '*') {
                        query += "\"codeOrNameLike\":\"" + searchValue.searchValue + "\",";
                    }
                    else {
                        query += "\"" + searchValue.searchField + "\":\"" + searchValue.searchValue + "\",";
                    }
                }
                if (query.substr(query.length - 1, 1) === ',') {
                    query = query.substring(0, query.length - 1);
                }
                query += '}';
                query = encodeURIComponent(query);
                query += "&pageIndex=" + index + "&pageSize=" + size;
                url += '?searchValue=' + query;
            }
            else {
                url += "?pageIndex=" + index + "&pageSize=" + size;
            }
            return this.http.get(url).pipe(map((/**
             * @param {?} re
             * @return {?}
             */
            function (re) { return ({
                items: re.content,
                total: re.totalElements,
                pageInfo: {
                    pageIndex: re.pageable.pageNumber + 1,
                    pageSize: re.pageable.pageSize,
                    pageList: [20, 40, 60, 80, 100],
                    enablePager: true
                }
            }); })));
        }
    };
    MatrixDefHelpService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MatrixDefHelpService.ctorParameters = function () { return [
        { type: HttpService }
    ]; };
    return MatrixDefHelpService;
}());
export { MatrixDefHelpService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MatrixDefHelpService.prototype.http;
}
