// @edp-aif/runtime-api/esm2015/lib/entity/rpc/RpcWritebackRuleExecutorArgs.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { MappingAssociation } from '../MappingAssociation';
export class RpcWritebackRuleExecutorArgs extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.sourceVoEntities = new Array();
        this.associations = new Array();
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.writebackRuleId = null;
        this.sourceVoId = null;
        this.targetVoFuncSessionId = null;
        this.sourceVoEntities = new Array();
        this.associations = new Array();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "writebackRuleId", this.writebackRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoFuncSessionId", this.targetVoFuncSessionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntities", BpmnModelJsonHelper.StringListToJson(this.sourceVoEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.writebackRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackRuleId");
        this.sourceVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
        this.targetVoFuncSessionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoFuncSessionId");
        this.sourceVoEntities = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoEntities"));
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
    }
}
if (false) {
    /** @type {?} */
    RpcWritebackRuleExecutorArgs.prototype.writebackRuleId;
    /** @type {?} */
    RpcWritebackRuleExecutorArgs.prototype.sourceVoId;
    /** @type {?} */
    RpcWritebackRuleExecutorArgs.prototype.sourceVoEntities;
    /** @type {?} */
    RpcWritebackRuleExecutorArgs.prototype.associations;
    /** @type {?} */
    RpcWritebackRuleExecutorArgs.prototype.targetVoFuncSessionId;
}
