// @farris/ui-datatable/esm2015/lib/datatable.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * @record
 */
export function SelectRowArgs() { }
if (false) {
    /** @type {?} */
    SelectRowArgs.prototype.rowIndex;
    /** @type {?} */
    SelectRowArgs.prototype.rowData;
}
export class DataTableService {
    constructor() {
        this.loadSuccess = new Subject();
        /**
         * 选中行事件
         */
        this.selectedRow = new Subject();
        this.unSelectedRow = new Subject();
        this.selectedAll = new Subject();
        this.dblClickRow = new Subject();
        this.updateCheckAllStatus = new Subject();
    }
}
DataTableService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DataTableService.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    DataTableService.prototype.loadSuccess;
    /**
     * 选中行事件
     * @type {?}
     */
    DataTableService.prototype.selectedRow;
    /** @type {?} */
    DataTableService.prototype.unSelectedRow;
    /** @type {?} */
    DataTableService.prototype.selectedAll;
    /** @type {?} */
    DataTableService.prototype.dblClickRow;
    /** @type {?} */
    DataTableService.prototype.updateCheckAllStatus;
}
