// @farris/ui-datatable/esm2015/lib/datatable-column.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, HostBinding } from '@angular/core';
export class ColumnDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        this.align = 'left';
        this.multipleFilter = true;
        this.fixedClass = this.fixed === 'left';
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
}
ColumnDirective.decorators = [
    { type: Directive, args: [{
                selector: 'data-column'
            },] }
];
/** @nocollapse */
ColumnDirective.ctorParameters = () => [
    { type: ElementRef }
];
ColumnDirective.propDecorators = {
    edit: [{ type: Input }],
    title: [{ type: Input }],
    field: [{ type: Input }],
    width: [{ type: Input }],
    align: [{ type: Input }],
    fixed: [{ type: Input }],
    className: [{ type: Input }],
    media: [{ type: Input }],
    sortable: [{ type: Input }],
    multipleFilter: [{ type: Input }],
    filter: [{ type: Input }],
    fixedClass: [{ type: HostBinding, args: ['class.fixed',] }]
};
if (false) {
    /** @type {?} */
    ColumnDirective.prototype.edit;
    /** @type {?} */
    ColumnDirective.prototype.title;
    /** @type {?} */
    ColumnDirective.prototype.field;
    /** @type {?} */
    ColumnDirective.prototype.width;
    /** @type {?} */
    ColumnDirective.prototype.align;
    /** @type {?} */
    ColumnDirective.prototype.fixed;
    /** @type {?} */
    ColumnDirective.prototype.className;
    /** @type {?} */
    ColumnDirective.prototype.media;
    /** @type {?} */
    ColumnDirective.prototype.sortable;
    /** @type {?} */
    ColumnDirective.prototype.multipleFilter;
    /** @type {?} */
    ColumnDirective.prototype.filter;
    /** @type {?} */
    ColumnDirective.prototype.fixedClass;
    /**
     * @type {?}
     * @private
     */
    ColumnDirective.prototype.el;
}
