// @farris/ui-datatable/esm2015/lib/datatable-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
export class DTCheckboxComponent {
    constructor() {
        this.checkedChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @param {?} event
     * @return {?}
     */
    handleClick(event) {
        if (!this.disabled) {
            this.checkedChange.emit({ originalEvent: event, checked: !this.checked });
        }
    }
}
DTCheckboxComponent.decorators = [
    { type: Component, args: [{
                selector: 'dt-checkbox',
                template: `
        <div class="custom-control custom-checkbox  custom-control-inline dt-checkbox f-checkradio-single">
            <input type="checkbox" class="custom-control-input"  [checked]="checked">
            <label class="custom-control-label" (click)="handleClick($event)"></label>
        </div>
    `
            }] }
];
/** @nocollapse */
DTCheckboxComponent.ctorParameters = () => [];
DTCheckboxComponent.propDecorators = {
    checked: [{ type: Input }],
    id: [{ type: Input }],
    disabled: [{ type: Input }],
    checkedChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DTCheckboxComponent.prototype.checked;
    /** @type {?} */
    DTCheckboxComponent.prototype.id;
    /** @type {?} */
    DTCheckboxComponent.prototype.disabled;
    /** @type {?} */
    DTCheckboxComponent.prototype.checkedChange;
}
