// @gsp-dip/data-imp-exp/esm5/lib/model/CommonStep/ActualParameter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 构件参数
 */
var /**
 * 构件参数
 */
ActualParameter = /** @class */ (function () {
    function ActualParameter() {
        /**
         * 参数内码
         */
        this.ID = '';
        /**
         * 参数编号
         */
        this.Code = '';
        /**
         * 参数名称
         */
        this.Name = '';
        /**
         * 类型数据类型
         */
        this.ValueType = ValueType.Constant;
        /**
         * 参数取值
         */
        this.Value = '';
        /**
         * 是否必须
         */
        this.IsRequired = true;
    }
    /**
     * copy
     */
    /**
     * copy
     * @param {?} paramter
     * @return {?}
     */
    ActualParameter.copy = /**
     * copy
     * @param {?} paramter
     * @return {?}
     */
    function (paramter) {
        /** @type {?} */
        var componentParameter = new ActualParameter();
        componentParameter.ID = paramter.ID;
        componentParameter.Code = paramter.Code;
        componentParameter.Name = paramter.Name;
        componentParameter.ValueType = paramter.ValueType;
        componentParameter.Value = paramter.Value;
        componentParameter.IsRequired = paramter.IsRequired;
        return componentParameter;
    };
    return ActualParameter;
}());
/**
 * 构件参数
 */
export { ActualParameter };
if (false) {
    /**
     * 参数内码
     * @type {?}
     */
    ActualParameter.prototype.ID;
    /**
     * 参数编号
     * @type {?}
     */
    ActualParameter.prototype.Code;
    /**
     * 参数名称
     * @type {?}
     */
    ActualParameter.prototype.Name;
    /**
     * 类型数据类型
     * @type {?}
     */
    ActualParameter.prototype.ValueType;
    /**
     * 参数取值
     * @type {?}
     */
    ActualParameter.prototype.Value;
    /**
     * 是否必须
     * @type {?}
     */
    ActualParameter.prototype.IsRequired;
}
/** @enum {number} */
var ValueType = {
    /**
     * 常量
     */
    Constant: 0,
    /**
     * 系统变量
     */
    SysVariable: 1,
};
export { ValueType };
ValueType[ValueType.Constant] = 'Constant';
ValueType[ValueType.SysVariable] = 'SysVariable';
/**
 * 参数取值类型Grid用
 * @type {?}
 */
export var ValueTypes = [
    {
        type: ValueType.Constant,
        name: '常量'
    },
    {
        type: ValueType.SysVariable,
        name: '系统变量'
    }
];
