// @gsp-dip/data-imp-exp/esm5/lib/model/CommonStep/CommonStep.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { IEStepType } from '../IEStepType';
import { Guid } from '../utils/Guid';
/**
 * 公共步骤实体
 */
var /**
 * 公共步骤实体
 */
CommonStep = /** @class */ (function () {
    function CommonStep() {
        /**
         * 内码
         */
        this.StepID = Guid.newGuid();
        /**
         *  步骤执行顺序
         */
        this.StortOrder = 1;
        /**
         * 步骤类型
         */
        this.StepType = IEStepType.Default;
        /**
         * 源模型Id
         */
        this.SrcModelId = '';
        /**
         * 目标模型Id
         */
        this.TarModelId = '';
        /**
         *  步骤名称
         */
        this.StepName = '';
        /**
         *  是否用户自定义
         */
        this.Custom = false;
        /**
         *  用户自定义构件标识
         */
        this.CustomComponent = '';
        /**
         *  步骤前事件列表
         */
        this.DoBefore = new Array();
        /**
         *  步骤后事件列表
         */
        this.DoAfter = new Array();
    }
    /**
     * copy
     */
    /**
     * copy
     * @param {?} step
     * @return {?}
     */
    CommonStep.copy = /**
     * copy
     * @param {?} step
     * @return {?}
     */
    function (step) {
        var e_1, _a, e_2, _b;
        /** @type {?} */
        var commonStep = new CommonStep();
        commonStep.StepID = step.StepID;
        commonStep.StortOrder = step.StortOrder;
        commonStep.StepType = step.StepType;
        commonStep.SrcModelId = step.SrcModelId;
        commonStep.TarModelId = step.TarModelId;
        commonStep.StepName = step.StepName;
        commonStep.Custom = step.Custom;
        commonStep.CustomComponent = step.CustomComponent;
        commonStep.DoBefore = new Array();
        try {
            for (var _c = tslib_1.__values(step.DoBefore), _d = _c.next(); !_d.done; _d = _c.next()) {
                var beforeCom = _d.value;
                commonStep.DoBefore.push(beforeCom);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        commonStep.DoAfter = new Array();
        try {
            for (var _e = tslib_1.__values(step.DoAfter), _f = _e.next(); !_f.done; _f = _e.next()) {
                var afterCom = _f.value;
                commonStep.DoAfter.push(afterCom);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return commonStep;
    };
    return CommonStep;
}());
/**
 * 公共步骤实体
 */
export { CommonStep };
if (false) {
    /**
     * 内码
     * @type {?}
     */
    CommonStep.prototype.StepID;
    /**
     *  步骤执行顺序
     * @type {?}
     */
    CommonStep.prototype.StortOrder;
    /**
     * 步骤类型
     * @type {?}
     */
    CommonStep.prototype.StepType;
    /**
     * 源模型Id
     * @type {?}
     */
    CommonStep.prototype.SrcModelId;
    /**
     * 目标模型Id
     * @type {?}
     */
    CommonStep.prototype.TarModelId;
    /**
     *  步骤名称
     * @type {?}
     */
    CommonStep.prototype.StepName;
    /**
     *  是否用户自定义
     * @type {?}
     */
    CommonStep.prototype.Custom;
    /**
     *  用户自定义构件标识
     * @type {?}
     */
    CommonStep.prototype.CustomComponent;
    /**
     *  步骤前事件列表
     * @type {?}
     */
    CommonStep.prototype.DoBefore;
    /**
     *  步骤后事件列表
     * @type {?}
     */
    CommonStep.prototype.DoAfter;
}
