// @qdp/common/esm2015/lib/dnd-list/services/dnd-draggable.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
export class DndDraggableService {
    constructor() {
        this._isDragging = false;
        this._itemType = '';
        this._removeOnDrop = false;
        this.dropEndSource = new Subject();
    }
    /**
     * @param {?} state
     * @return {?}
     */
    setDraggingState(state) {
        this._isDragging = state;
    }
    /**
     * @return {?}
     */
    getDraggingState() {
        return this._isDragging;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setItemType(value) {
        this._itemType = value.toLowerCase();
    }
    /**
     * @return {?}
     */
    getItemType() {
        return this._itemType;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setDropCallback(value) {
        this._dropCallback = value;
    }
    /**
     * @return {?}
     */
    getDropCallback() {
        return this._dropCallback;
    }
    /**
     * @param {?} dropEffect
     * @return {?}
     */
    setDropEffect(dropEffect) {
        this._dropEffect = dropEffect;
    }
    /**
     * @return {?}
     */
    getDropEffect() {
        return this._dropEffect;
    }
    /**
     * @param {?} effectAllowed
     * @return {?}
     */
    setEffectAllowed(effectAllowed) {
        this._effectAllowed = effectAllowed;
    }
    /**
     * @return {?}
     */
    getEffectAllowed() {
        return this._effectAllowed;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setRemoveOnDrop(value) {
        this._removeOnDrop = value;
    }
    /**
     * @return {?}
     */
    getRemoveOnDrop() {
        return this._removeOnDrop;
    }
}
DndDraggableService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DndDraggableService.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    DndDraggableService.prototype._isDragging;
    /**
     * @type {?}
     * @private
     */
    DndDraggableService.prototype._itemType;
    /**
     * @type {?}
     * @private
     */
    DndDraggableService.prototype._dropCallback;
    /**
     * @type {?}
     * @private
     */
    DndDraggableService.prototype._dropEffect;
    /**
     * @type {?}
     * @private
     */
    DndDraggableService.prototype._effectAllowed;
    /**
     * @type {?}
     * @private
     */
    DndDraggableService.prototype._removeOnDrop;
    /** @type {?} */
    DndDraggableService.prototype.dropEndSource;
}
