// @farris/ui-filter-panel/esm2015/lib/components/datecomponent/date.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
export class PanelDateComponent {
    constructor() {
        this.searchHandle = new EventEmitter();
        this.submitChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get item() {
        return this._item;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    set item(data) {
        this._item = data;
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.initDate();
        }), 0);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    initDate() {
        if (this._item.placeHolder) {
            this.placeholder = this._item.placeHolder || '';
        }
        if (!this._item.hasOwnProperty('value')) {
            this._item.value = { value: '' };
        }
        if (this.item && this._item.value
            && (this._item.value).value) {
            this.date = this._item.value.value;
        }
        else {
            this.date = undefined;
        }
        /** @type {?} */
        const format = this.showType == '2' ? 'yyyy-MM' : (this.showType == '3' ? 'yyyy' : 'yyyy-MM-dd');
        this.dateFormat = this._item.control.format ? this._item.control.format : format;
        this.setSubmitChange();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onDateChange(event) {
        if (event.returnFormatted) {
            (this._item.value).value = event.returnFormatted;
        }
        else {
            (this._item.value).value = '';
        }
        // this._item.valueText = this._item.value.value;
        this._item.valueText = event.formatted;
        this.setSubmitChange();
        this.searchHandle.emit(event);
    }
    /**
     * @return {?}
     */
    setSubmitChange() {
        /** @type {?} */
        const val = this._item.value;
        if (Object.keys(val).length < 0 || !val.value) {
            this.submitChange.emit(false);
        }
        else {
            this.submitChange.emit(true);
        }
    }
}
PanelDateComponent.decorators = [
    { type: Component, args: [{
                selector: 'panel-filter-date',
                template: "\r\n<farris-datepicker\r\n    [(ngModel)]=\"date\"\r\n    [returnFormat]=\"showType == '2'? 'yyyy-MM': (showType == '3'? 'yyyy':'yyyy-MM-dd')\"\r\n    [editable]=\"false\" [dateFormat]=\"dateFormat\"\r\n    (valueChange)=\"onDateChange($event)\"\r\n    [placeholder]=\"placeholder\" [showType]=\"showType\">\r\n</farris-datepicker>"
            }] }
];
/** @nocollapse */
PanelDateComponent.ctorParameters = () => [];
PanelDateComponent.propDecorators = {
    item: [{ type: Input }],
    showType: [{ type: Input }],
    searchHandle: [{ type: Output }],
    submitChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    PanelDateComponent.prototype._item;
    /** @type {?} */
    PanelDateComponent.prototype.showType;
    /** @type {?} */
    PanelDateComponent.prototype.searchHandle;
    /** @type {?} */
    PanelDateComponent.prototype.submitChange;
    /** @type {?} */
    PanelDateComponent.prototype.date;
    /** @type {?} */
    PanelDateComponent.prototype.placeholder;
    /** @type {?} */
    PanelDateComponent.prototype.dateFormat;
}
