// @gsp-cmp/querysolution/esm2015/lib/saveasdialog/saveasdialog.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/saveasdialog/saveasdialog.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, ViewChild, HostBinding, ViewEncapsulation, Input } from '@angular/core';
import { NotifyService } from '@farris/ui-notify';
import { QuerySolutionVo, QuerySolutionExtendVo, SolutionService } from '@farris/component-querycondition';
import { LocaleService } from '@farris/ui-locale';
import { DialogComponent } from '@farris/ui-dialog';
import { cloneDeep } from 'lodash-es';
export class SaveAsDialogComponent {
    /**
     * @param {?} notify
     * @param {?} solutionService
     * @param {?} localeService
     */
    constructor(notify, solutionService, localeService) {
        this.notify = notify;
        this.solutionService = solutionService;
        this.localeService = localeService;
        this.cls = 'saveasdialog';
        this.cancelEmitter = new EventEmitter();
        this.confirmEmitter = new EventEmitter();
        this.getAuthority = new EventEmitter();
        //@Input属性，旧代码未设置，故保留
        this.querySolution = new QuerySolutionVo();
        // 标记code中是否有异常
        this.codeError = false;
        // code最长是100个字符，不论中英文
        this.codeMaxLength = 100;
        // 方案类型：用户个人、全局公共、组织公共
        this.solutionType = [{
                value: 'private',
                name: this.localeService.getValue("extendQuerySolution.private.setAsPrivate")
            },
            {
                value: 'public',
                name: this.localeService.getValue("extendQuerySolution.public.setAsPublic")
            },
            {
                value: 'org',
                name: this.localeService.getValue("extendQuerySolution.org.setAsOrgPublic")
            }];
        // 当前已选方案类型
        this.radioGroupValue = 'private';
        // 是否已分配功能权限，未分配时值为true
        this.authorityError = false;
        // 是否已分配数据权限，未分配时值为true 
        this.noDataAuthority = false;
        // 分配的组织数据权限
        this.dataPermissionIds = [];
        // 组织选择组件的返回值
        this.selectedOrgIds = [];
        this.isDefault = false;
        this.languages = [
            { code: 'zh-cn', name: '中文简体', default: true },
            { code: 'en', name: 'English' },
            { code: 'zh-CHT', name: '中文繁体' }
        ];
    }
    /**
     * @param {?} e
     * @return {?}
     */
    set hideOrgselector(e) {
        //隐藏组织方案
        if (e) {
            this.solutionType = this.solutionType.filter((/**
             * @param {?} val
             * @return {?}
             */
            val => val.value !== 'org'));
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.pRadioGroupValue = this.radioGroupValue;
        this.placeholder = this.localeService.getValue('extendQuerySolution.org.placeholder');
        //兼容旧版设计器
        this.currentLang = ((/** @type {?} */ (window))).gspframeworkService ? ((/** @type {?} */ (window))).gspframeworkService.rtf.language.getLanguageCode() : 'zh-cn';
    }
    /**
     * @return {?}
     */
    show() {
        this.dialog.show();
    }
    /**
     * @return {?}
     */
    cancel() {
        // this.cancelEmitter.emit(true);
        this.solutionCodeObj = null;
        this.dialog.close();
    }
    /**
     * @return {?}
     */
    confirm() {
        if (this.solutionCodeObj) {
            this.codeChangeHandler(this.solutionCodeObj['zh-cn']);
            this.codeChangeHandler(this.solutionCodeObj['en']);
            this.codeChangeHandler(this.solutionCodeObj['zh-CHT']);
            if (this.codeError) {
                return;
            }
            this.querySolution.code_chs = this.solutionCodeObj['zh-cn'];
            this.querySolution.code_en = this.solutionCodeObj['en'];
            this.querySolution.code_cht = this.solutionCodeObj['zh-CHT'];
        }
        // 如果有错误返回
        // if (!this.querySolution.code) {
        //   this.solutionService.showInfo(this.localeService.getValue('querySolution.saveAsDialog.pleaseInput'), 'info', '', 1500);
        //   return;
        // }
        if (this.currentLang == 'zh-CHS') {
            if (!this.querySolution.code_chs) {
                this.solutionService.showInfo(this.localeService.getValue('querySolution.saveAsDialog.pleaseInput'), 'info', '', 1500);
                return;
            }
            this.querySolution.code = this.querySolution.code_chs;
        }
        else if (this.currentLang == 'en') {
            if (!this.querySolution.code_en) {
                this.solutionService.showInfo(this.localeService.getValue('querySolution.saveAsDialog.pleaseInput'), 'info', '', 1500);
                return;
            }
            this.querySolution.code = this.querySolution.code_en;
        }
        else if (this.currentLang == 'zh-CHT') {
            if (!this.querySolution.code_cht) {
                this.solutionService.showInfo(this.localeService.getValue('querySolution.saveAsDialog.pleaseInput'), 'info', '', 1500);
                return;
            }
            this.querySolution.code = this.querySolution.code_cht;
        }
        if (this.authorityError && this.radioGroupValue != 'private') {
            this.solutionService.showInfo(this.localeService.getValue('extendQuerySolution.auth.createError'), 'warning', '', 1500);
            return;
        }
        this.querySolution.type = this.radioGroupValue;
        this.querySolution.isDefault = this.radioGroupValue == 'private' ? this.isDefault : false;
        if (this.radioGroupValue == 'org') {
            if (this.selectedOrgIds.length) {
                this.querySolution.extendInfo = this.selectedOrgIds.map((/**
                 * @param {?} id
                 * @return {?}
                 */
                id => {
                    /** @type {?} */
                    let info = new QuerySolutionExtendVo();
                    info['id'] = this.solutionService.getGuid(),
                        info['relatedId'] = this.querySolution.id;
                    info['orgId'] = id;
                    return info;
                }));
            }
            else {
                this.solutionService.showInfo(this.localeService.getValue('extendQuerySolution.auth.noSelectedOrg'), 'info', '', 1500);
                return;
            }
        }
        /** @type {?} */
        let obj = new Object();
        obj['querySolution'] = cloneDeep(this.querySolution);
        this.confirmEmitter.emit(obj);
        this.solutionCodeObj = null;
        this.dialog.close();
    }
    /**
     * @param {?} checkStatus
     * @return {?}
     */
    setDefault(checkStatus) {
        //this.querySolution.isDefault = checkStatus;
        this.isDefault = checkStatus;
    }
    // 变化时检测
    /**
     * @param {?} result
     * @return {?}
     */
    codeChangeHandler(result) {
        if (result && result.length > this.codeMaxLength) {
            this.codeError = true;
        }
        else {
            this.codeError = false;
        }
    }
    // 切换方案类型：个人/公共/组织公共，需要判断权限
    /**
     * @param {?} e
     * @return {?}
     */
    changeSolutionType(e) {
        this.querySolution.type = e;
        this.querySolution.isDefault = e == 'private' ? this.isDefault : false;
        this.getAuthority.emit(e);
    }
    // 获取组织选择组件的返回值（1、弹窗内点确定；2、文本框内移除单个）
    /**
     * @param {?} e
     * @return {?}
     */
    getSelectedId(e) {
        this.selectedOrgIds = ((/** @type {?} */ (e.selections))).map((/**
         * @param {?} org
         * @return {?}
         */
        org => org.orgId));
    }
    // 清空文本框
    /**
     * @return {?}
     */
    clearAllOrgs() {
        this.selectedOrgIds = [];
    }
}
SaveAsDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-saveasdialog',
                template: "<farris-dialog #dialog [buttons]=\"saveasbuttonRef\" [width]=\"550\" [height]=\"310\" [title]=\"title\">\r\n  <div class=\"qs-save-as-input\">\r\n    <div class=\"farris-input-wrap\">\r\n      <!-- <input class=\"form-control\"  placeholder=\"{{'querySolution.saveAsDialog.placeholder' | locale}}\"\r\n        [(ngModel)]=\"querySolution.code\" (ngModelChange)=\"codeChangeHandler($event)\" /> -->\r\n      <language-textbox [editable]=\"true\" [languages]=\"languages\" [panelWidth]=\"530\" [(ngModel)]=\"solutionCodeObj\"></language-textbox>\r\n        <div class=\"farris-feedback\" [class.f-state-invalid]=\"codeError\">\r\n            <span class=\"f-feedback-message ng-star-inserted\">\u540D\u79F0\u957F\u5EA6\u4E0D\u5E94\u8D85\u8FC7100</span>\r\n        </div>\r\n    </div>\r\n    <div class=\"custom-control custom-checkbox qs-save-as-check\">\r\n      <farris-radiogroup [(ngModel)]=\"radioGroupValue\" [data]=\"solutionType\" [horizontal]=\"true\"\r\n        (changeValue)=\"changeSolutionType($event)\">\r\n      </farris-radiogroup>\r\n    </div>\r\n    <div class=\"custom-control custom-checkbox qs-save-as-check\" *ngIf=\"radioGroupValue == 'private'\">\r\n      <input class=\"custom-control-input\" type=\"checkbox\" id=\"defaultSolution\" (change)=\"setDefault($event.target.checked)\"\r\n        [(ngModel)]=\"isDefault\" />\r\n      <label class=\"custom-control-label\" id=\"defaultSolution\"\r\n        (click)=\"setDefault(!isDefault)\">{{'querySolution.saveAsDialog.setAsDefault' | locale}}</label>\r\n    </div>\r\n    <div class=\"custom-control custom-checkbox qs-save-as-check\" *ngIf=\"radioGroupValue == 'org' && !noDataAuthority\">\r\n      <farris-organization-selector\r\n        [primaryField]=\"'orgId'\" [displayField]=\"'name'\" [absOrgType]=\"'System_organization'\" [placeholder]=\"placeholder\"\r\n        [layer]=\"1\" [multiSelect]=\"true\" [showTabIds]=\"'organization'\" [enableHierarchicalLoading]=\"true\"\r\n        [cascadeCheck]=\"true\" [cascadeDown]=\"true\" [cascadeUp]=\"false\" [orgIds]=\"selectedOrgIds\"\r\n        [dataPermission]=\"true\" [dataPermissionIds]=\"dataPermissionIds\" (selectionsChange)=\"getSelectedId($event)\"\r\n        (inputClear)=\"clearAllOrgs()\" (tagRemoved)=\"getSelectedId($event)\"\r\n        ></farris-organization-selector>\r\n    </div>  \r\n  </div>\r\n</farris-dialog>\r\n<ng-template #saveasbuttonRef>\r\n  <div class=\"qs-save-as-btn\">\r\n    <button class=\"btn btn-primary\" [disabled]=\"codeError\"  (click)=\"confirm()\">{{'querySolution.saveAsDialog.confirm' | locale}}</button>\r\n    <button class=\"btn btn-secondary\" (click)=\"cancel()\">{{'querySolution.saveAsDialog.cancel' | locale}}</button>\r\n  </div>\r\n</ng-template>\r\n",
                encapsulation: ViewEncapsulation.None,
                styles: [".qs-save-as-btn .btn{float:right;margin-left:10px;font-size:14px;line-height:26px}.qs-save-as-btn .btn-secondary{min-width:70px}.qs-save-as-input{width:100%;line-height:20px;font-size:14px;padding:20px 10px}.qs-save-as-check{margin:10px auto}"]
            }] }
];
/** @nocollapse */
SaveAsDialogComponent.ctorParameters = () => [
    { type: NotifyService },
    { type: SolutionService },
    { type: LocaleService }
];
SaveAsDialogComponent.propDecorators = {
    cls: [{ type: HostBinding, args: ['class',] }],
    cancelEmitter: [{ type: Output }],
    confirmEmitter: [{ type: Output }],
    getAuthority: [{ type: Output }],
    hideOrgselector: [{ type: Input }],
    dialog: [{ type: ViewChild, args: ['dialog',] }]
};
if (false) {
    /** @type {?} */
    SaveAsDialogComponent.prototype.cls;
    /** @type {?} */
    SaveAsDialogComponent.prototype.cancelEmitter;
    /** @type {?} */
    SaveAsDialogComponent.prototype.confirmEmitter;
    /** @type {?} */
    SaveAsDialogComponent.prototype.getAuthority;
    /** @type {?} */
    SaveAsDialogComponent.prototype.querySolution;
    /** @type {?} */
    SaveAsDialogComponent.prototype.title;
    /** @type {?} */
    SaveAsDialogComponent.prototype.defaultQuerySolution;
    /** @type {?} */
    SaveAsDialogComponent.prototype.dialog;
    /** @type {?} */
    SaveAsDialogComponent.prototype.codeError;
    /** @type {?} */
    SaveAsDialogComponent.prototype.codeMaxLength;
    /** @type {?} */
    SaveAsDialogComponent.prototype.solutionType;
    /** @type {?} */
    SaveAsDialogComponent.prototype.radioGroupValue;
    /** @type {?} */
    SaveAsDialogComponent.prototype.pRadioGroupValue;
    /** @type {?} */
    SaveAsDialogComponent.prototype.authorityError;
    /** @type {?} */
    SaveAsDialogComponent.prototype.noDataAuthority;
    /** @type {?} */
    SaveAsDialogComponent.prototype.dataPermissionIds;
    /** @type {?} */
    SaveAsDialogComponent.prototype.selectedOrgIds;
    /** @type {?} */
    SaveAsDialogComponent.prototype.placeholder;
    /** @type {?} */
    SaveAsDialogComponent.prototype.isDefault;
    /**
     * @type {?}
     * @private
     */
    SaveAsDialogComponent.prototype.currentLang;
    /** @type {?} */
    SaveAsDialogComponent.prototype.languages;
    /** @type {?} */
    SaveAsDialogComponent.prototype.solutionCodeObj;
    /**
     * @type {?}
     * @private
     */
    SaveAsDialogComponent.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    SaveAsDialogComponent.prototype.solutionService;
    /**
     * @type {?}
     * @private
     */
    SaveAsDialogComponent.prototype.localeService;
}
