// @farris/ui-datepicker/esm5/lib/components/multi-day-view/multi-days.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
var MultiSelectedDaysComponent = /** @class */ (function () {
    function MultiSelectedDaysComponent() {
        this.locales = {};
        this.selectedDates = [];
        this.removed = new EventEmitter();
        this.clear = new EventEmitter();
        this.clicked = new EventEmitter();
    }
    /**
     * @return {?}
     */
    MultiSelectedDaysComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    MultiSelectedDaysComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
    };
    /**
     * @param {?} $event
     * @param {?} item
     * @return {?}
     */
    MultiSelectedDaysComponent.prototype.onRemove = /**
     * @param {?} $event
     * @param {?} item
     * @return {?}
     */
    function ($event, item) {
        $event.stopPropagation();
        this.removed.emit(item);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    MultiSelectedDaysComponent.prototype.onClear = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        $event.stopPropagation();
        this.clear.emit();
    };
    /**
     * @param {?} $event
     * @param {?} item
     * @return {?}
     */
    MultiSelectedDaysComponent.prototype.onSelectDateClick = /**
     * @param {?} $event
     * @param {?} item
     * @return {?}
     */
    function ($event, item) {
        $event.stopPropagation();
        this.clicked.emit(item);
    };
    MultiSelectedDaysComponent.decorators = [
        { type: Component, args: [{
                    selector: 'multi-days',
                    template: "\n        <div class=\"d-flex mb-2 f-datepicker-multi-selectdays-header\">\n            <span style=\"font-size: 14px;\" class=\"ml-3\">{{ locales?.selected.split(',')[0] || '\u5DF2\u9009'}} <span style=\"font-weight: bold;font-size: 17px;\">{{selectedDates.length}}</span> {{locales?.selected.split(',')[1] ||'\u5929'}}</span>\n            <span class=\"f-datepicker-multi-selectdays-clear\" (click)=\"onClear($event)\">{{ locales?.clearSelections || '\u6E05\u7A7A'}}</span>\n        </div>\n        <div #selecteditems style=\"height: 230px;overflow-y: auto;\" class=\"ml-3 multi-date-select win11Scroll nobtn\" (wheel)=\"$event.stopPropagation()\">\n            <ul [style.width.px]=\"167\">\n                <li class=\"d-flex align-items-center f-datepicker-multi-selectdays\" *ngFor=\"let d of selectedDates\" (click)=\"onSelectDateClick($event, d)\">\n                    {{d.value}} <span class=\"f-icon f-icon-close remove\" title=\"{{ locales?.delete || '\u5220\u9664'}}\" (click)=\"onRemove($event, d)\"></span>\n                </li>\n            </ul>\n        </div>\n    ",
                    styles: ["\n        .f-datepicker-multi-selectdays-header{\n            height: 41px;border-bottom: 1px solid #E4E7EF;line-height: 40px;justify-content: space-between;\n        }\n\n        .f-datepicker-multi-selectdays {\n            justify-content: space-between;color: #2D2F33;height: 30px;padding: 0 5px;border-radius: 6px;\n            cursor: pointer;\n        }\n\n        .f-datepicker-multi-selectdays:hover {\n            background: #F7F8FB;\n        }\n\n        .f-datepicker-multi-selectdays .remove {\n            font-size: 12px;color: #C7CFDD;cursor: pointer;\n        }\n\n        .f-datepicker-multi-selectdays .remove:hover {\n            color: #5f96f5;\n        }\n\n        .f-datepicker-multi-selectdays-clear {\n            font-size: 13px;margin-right: 10px; color: #CCD1DD; cursor: pointer;\n        }\n        .f-datepicker-multi-selectdays-clear:hover {\n            color: #5f96f5;\n        }\n\n        .mr14 {\n            margin-right: 14px;\n        }\n        "]
                }] }
    ];
    /** @nocollapse */
    MultiSelectedDaysComponent.ctorParameters = function () { return []; };
    MultiSelectedDaysComponent.propDecorators = {
        locales: [{ type: Input }],
        selectedDates: [{ type: Input }],
        removed: [{ type: Output }],
        clear: [{ type: Output }],
        clicked: [{ type: Output }],
        selecteditemsEl: [{ type: ViewChild, args: ['selecteditems',] }]
    };
    return MultiSelectedDaysComponent;
}());
export { MultiSelectedDaysComponent };
if (false) {
    /** @type {?} */
    MultiSelectedDaysComponent.prototype.locales;
    /** @type {?} */
    MultiSelectedDaysComponent.prototype.selectedDates;
    /** @type {?} */
    MultiSelectedDaysComponent.prototype.removed;
    /** @type {?} */
    MultiSelectedDaysComponent.prototype.clear;
    /** @type {?} */
    MultiSelectedDaysComponent.prototype.clicked;
    /** @type {?} */
    MultiSelectedDaysComponent.prototype.selecteditemsEl;
}
