// @farris/header-group-editor/esm5/lib/group-header-select-columns.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectorRef } from '@angular/core';
var SelectColumnsComponent = /** @class */ (function () {
    function SelectColumnsComponent(cd) {
        this.cd = cd;
        this.fields = [];
        this.checkeds = [];
        this.uncheckeds = [];
        this.childs = [];
        this.checkedPropertyName = '__checked__';
        // 候选列
        this.hxl = [];
    }
    /**
     * @return {?}
     */
    SelectColumnsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initCheckedProperty();
        this.childs = Array.from(this.checkeds);
        if (this.groupEditorRef) {
            this.hxl = this.groupEditorRef.getHouXuanLie();
            if (this.hxl.length) {
                this.hxl = this.hxl.map((/**
                 * @param {?} n
                 * @return {?}
                 */
                function (n) { return n.data; }));
            }
        }
    };
    /**
     * @private
     * @return {?}
     */
    SelectColumnsComponent.prototype.initCheckedProperty = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.checkeds) {
            this.checkeds = [];
            this.fields.forEach((/**
             * @param {?} f
             * @return {?}
             */
            function (f) {
                f[_this.checkedPropertyName] = false;
            }));
        }
        else {
            this.fields.forEach((/**
             * @param {?} f
             * @return {?}
             */
            function (f) {
                f[_this.checkedPropertyName] = _this.isChecked(f);
            }));
        }
    };
    /**
     * @param {?} f
     * @return {?}
     */
    SelectColumnsComponent.prototype.checkedCols = /**
     * @param {?} f
     * @return {?}
     */
    function (f) {
        var _this = this;
        if (this.isDisabled(f)) {
            return;
        }
        /** @type {?} */
        var checked = this.isChecked(f);
        f[this.checkedPropertyName] = !checked;
        /** @type {?} */
        var _fields = this.fields.filter((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return !_this.isDisabled(n); }));
        this.checkeds = _fields.filter((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n[_this.checkedPropertyName]; }));
        this.uncheckeds = _fields.filter((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return !n[_this.checkedPropertyName]; }));
        this.cd.detectChanges();
    };
    /**
     * @param {?} f
     * @return {?}
     */
    SelectColumnsComponent.prototype.isChecked = /**
     * @param {?} f
     * @return {?}
     */
    function (f) {
        /** @type {?} */
        var chk = false;
        if (this.checkeds && this.checkeds.length) {
            chk = this.checkeds.findIndex((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return n.id === f.id; })) > -1;
        }
        return chk;
    };
    /** 在其他分组中的列，禁用选择
     * 仅当前分组中的列及候选列中的项可以选择
     */
    /**
     * 在其他分组中的列，禁用选择
     * 仅当前分组中的列及候选列中的项可以选择
     * @param {?} f
     * @return {?}
     */
    SelectColumnsComponent.prototype.isDisabled = /**
     * 在其他分组中的列，禁用选择
     * 仅当前分组中的列及候选列中的项可以选择
     * @param {?} f
     * @return {?}
     */
    function (f) {
        /** @type {?} */
        var items = this.childs;
        if (this.hxl && this.hxl.length) {
            items = items.concat(this.hxl);
        }
        return !(items.findIndex((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n.id === f.id; })) > -1);
    };
    SelectColumnsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'select-columns',
                    template: "\n    <div style=\"margin: 10px 20px\" class=\"row\">\n        <div class=\"custom-control custom-checkbox col-4\" *ngFor=\"let f of fields\">\n            <input type=\"checkbox\" class=\"custom-control-input\" [checked]=\"isChecked(f)\" [disabled]=\"isDisabled(f)\">\n            <label class=\"custom-control-label\" (click)=\"checkedCols(f)\" >{{ f.caption }}</label>\n        </div>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    SelectColumnsComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    SelectColumnsComponent.propDecorators = {
        fields: [{ type: Input }]
    };
    return SelectColumnsComponent;
}());
export { SelectColumnsComponent };
if (false) {
    /** @type {?} */
    SelectColumnsComponent.prototype.fields;
    /** @type {?} */
    SelectColumnsComponent.prototype.checkeds;
    /** @type {?} */
    SelectColumnsComponent.prototype.uncheckeds;
    /** @type {?} */
    SelectColumnsComponent.prototype.childs;
    /** @type {?} */
    SelectColumnsComponent.prototype.groupEditorRef;
    /**
     * @type {?}
     * @private
     */
    SelectColumnsComponent.prototype.checkedPropertyName;
    /**
     * @type {?}
     * @private
     */
    SelectColumnsComponent.prototype.hxl;
    /**
     * @type {?}
     * @private
     */
    SelectColumnsComponent.prototype.cd;
}
