// @farris/ui-adminorganization-selector/esm5/lib/organization-selector.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/organization-selector.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, forwardRef, HostBinding, Injector, Input, NgZone, Output, Renderer2, ViewChild, ChangeDetectorRef } from '@angular/core';
import { NgControl, NG_VALUE_ACCESSOR } from '@angular/forms';
import { DialogComponent } from '@farris/ui-dialog';
import { ApiService } from '../utils/api.service';
import { CommonUtil } from '../utils/common.tool';
import { FavoritesTool } from '../utils/favorites.tool';
import { ChangeType } from '../utils/index.type';
import { StateMgrTool } from '../utils/state.mgr.tool';
import { Subject } from 'rxjs';
import * as _ from 'lodash-es';
import { debounceTime, distinctUntilChanged, switchMap, tap } from 'rxjs/operators';
import { GroupComponent } from './tab-group/group.component';
import { OrganizationComponent } from './tab-organization/to.component';
import { FavoritesComponent } from './tab-favorites/favorites.component';
import { MessagerService } from '@farris/ui-messager';
import { NotifyService } from '@farris/ui-notify';
import { LocaleService } from '@farris/ui-locale';
import { ORGANIZATION_SELECTOR_LOCALE } from '../utils/locale';
var AdminOrganizationSelectorComponent = /** @class */ (function () {
    function AdminOrganizationSelectorComponent(render, injector, favoritesTool, apiService, ngZone, stateMgr) {
        var _this = this;
        this.render = render;
        this.injector = injector;
        this.favoritesTool = favoritesTool;
        this.apiService = apiService;
        this.ngZone = ngZone;
        this.stateMgr = stateMgr;
        this.baseCls = true;
        this.width = '100%';
        // 是否为零代码轻应用表格模板
        this.isTableTemplate = false;
        // input框是否显示
        this.showInput = true;
        // 业务标识
        this.busNum = 'common';
        this.multiSelect = true;
        this.placeholder = '请选择';
        this.searchPlaceholder = '请输入编号或名称搜索';
        this.readonly = false;
        // 传入的url
        this.absUserOrgUrl = '/api/runtime/sys/v1.0/absUserOrg';
        this.absRfQueryUrl = '/api/runtime/sys/v1.0/absRecentContacts/query';
        this.absRfModifyUrl = '/api/runtime/sys/v1.0/absRecentContacts/modify';
        this.groupApiUrl = "/api/runtime/sys/v1.0/org-group";
        this.groupExtendUrl = '/platform/runtime/wf/webapp/ng-farris-mobile-ui/extend-org.js';
        /**
         * 数据源类型
         * 默认为： System_organization  系统数据源   获取系统人员、系统组织
         * Administrative_organization  已扩展数据源  获取行政组织、行政人员
         * absOrgType {string}
         * PersonnelSelectorComponent
         */
        this.absOrgType = 'System_organization';
        // 启用分层加载
        this.enableHierarchicalLoading = true;
        // 不包含停用组织
        this.notGetStopOrg = true;
        this.primaryField = 'orgId';
        this.displayField = 'name';
        this.orgIds = [];
        // 展开层级
        this.layer = 1;
        // 是否启用级联选中
        this.cascadeCheck = true;
        // 级联选中下级
        this.cascadeDown = true;
        // 级联选中上级
        this.cascadeUp = false;
        // 显示全选按钮
        this.showCheckAll = false;
        this.keepSelect = false;
        // 一次性最多选中多少条
        this.selectedLimit = 50;
        this.showTabIds = 'organization,favorites';
        this.activeTabId = 'organization';
        // 搜索前端分页50条
        this.searchPageSize = 50;
        /**
         * 多选分隔符
         */
        this.multipleChoiceSeparator = ',';
        this.draggable = true;
        this.virtualizedLoad = false;
        this.dataPermission = false;
        this.dataPermissionIds = [];
        this.orgNameType = 'short';
        this.onlyDisplayCompany = false;
        this.onlyParentFinalText = false;
        this.quickSelect = false;
        this.selectionsChange = new EventEmitter();
        // 全部清除按钮触发回调
        this.inputClear = new EventEmitter();
        // 标签删除按钮出发回调
        this.tagRemoved = new EventEmitter();
        this.selections = [];
        this.originalSelections = [];
        this.finalText = [];
        this.parentFinal = [];
        this.abbreParentFinal = [];
        this.lookupIcon = '<i class="f-icon f-icon-lookup"></i>';
        this.searchIcon = '<span class="k-icon k-i-search"></span>';
        this.organizationTabId = 'organization';
        this.groupTabId = 'group';
        this.favoritesTabId = 'favorites';
        this.dialogHeight = 548;
        this.dialogId = 'dialogId';
        this.isShow = false;
        this.confirmToken = false;
        this.hideSearchPanel = true;
        this.searchText = '';
        this.groupDialogOpend = false;
        this.groupDialogTitle = '';
        this.validGroupTitle = true;
        this.groupTitleStream = new Subject();
        this.selectionsShow = false;
        this.orgTabActive = false;
        this.groupTabActive = false;
        this.againstSingleSelect = false;
        this.localeService = null;
        this.languagesArray = [
            { code: 'zh-CHS', name: '中文简体', default: true },
            { code: 'en', name: 'English' },
            { code: 'zh-CHT', name: '中文繁体' }
        ];
        this._onlyCurrentAdminOrg = false;
        this.farrisDialogbeforeClose = (/**
         * @return {?}
         */
        function () {
            _this.isShow = false;
            if (_this.confirmToken) {
                _this.originalSelections = _.cloneDeep(_this.selections);
            }
            else {
                _this.selections = _.cloneDeep(_this.originalSelections);
            }
            if (!_this.hideSearchPanel || _this.searchText) {
                _this.searchText = null;
                _this.clearSearch();
            }
            _this.confirmToken = false;
            return true;
        });
        this.onModelChange = (/**
         * @param {?=} val
         * @return {?}
         */
        function (val) { });
        this.onModelTouched = (/**
         * @param {?=} val
         * @return {?}
         */
        function (val) { });
        this.apiService.absUrl = this.absUserOrgUrl;
        this.apiService.groupApiUrl = this.groupApiUrl;
        this.apiService.getOrgsUrl = '/api/runtime/sys/v1.0/adminOrg';
        this.cdRef = this.injector.get(ChangeDetectorRef);
        this.messageService = this.injector.get(MessagerService, null);
        this.notifyService = this.injector.get(NotifyService, null);
        this.localeService = this.injector.get(LocaleService);
        if (!this.localeService.langData['organizationSelector']) {
            this.localeService.setLocaleData(ORGANIZATION_SELECTOR_LOCALE);
        }
    }
    Object.defineProperty(AdminOrganizationSelectorComponent.prototype, "bableTemplClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.isTableTemplate;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AdminOrganizationSelectorComponent.prototype, "onlyCurrentAdminOrg", {
        get: /**
         * @return {?}
         */
        function () {
            return this._onlyCurrentAdminOrg;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._onlyCurrentAdminOrg = this.onlyDisplayCompany ? false : value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} obj
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.writeValue = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        if (!!obj && typeof obj === 'string') {
            this.finalText = obj.split(this.multipleChoiceSeparator);
        }
        else {
            this.finalText = [];
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onModelChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onModelTouched = fn;
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.setDisabledState = /**
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
        throw new Error('Method not implemented.');
    };
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initDialogHeight();
        this.listenSelectionsChange();
        this.initTabs();
        this.initGroupSearch();
        this.ngControl = this.injector.get(NgControl, null);
        this.mapFields && this.mapFields2Component();
    };
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.listenWindowResize();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        /** @type {?} */
        var orgIdsChange = changes.orgIds;
        if (orgIdsChange && orgIdsChange.currentValue) {
            this.assembleDataByOrgIds(orgIdsChange.currentValue);
        }
    };
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.selectionsSub.unsubscribe();
    };
    /**
     * @private
     * @param {?} selections
     * @param {?=} changeType
     * @param {?=} from
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.emitSelections = /**
     * @private
     * @param {?} selections
     * @param {?=} changeType
     * @param {?=} from
     * @return {?}
     */
    function (selections, changeType, from) {
        if (changeType === void 0) { changeType = ChangeType.ValueChanged; }
        if (from === void 0) { from = 'form'; }
        this.stateMgr.emitSelections(selections, changeType, from);
    };
    // 根据字段映射关系获取当前已选组织的id，进而在打开选组织弹窗时进行相应展示
    // 根据字段映射关系获取当前已选组织的id，进而在打开选组织弹窗时进行相应展示
    /**
     * @private
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.mapFields2Component = 
    // 根据字段映射关系获取当前已选组织的id，进而在打开选组织弹窗时进行相应展示
    /**
     * @private
     * @return {?}
     */
    function () {
        // const helpFields = Object.keys(this.mapFields);
        // if (!helpFields.includes(this.primaryField)) {
        //   // this.notifyService.warning(this.localeService.getValue("organizationSelector.errorMessage.mapFields"));
        //   if(!helpFields.includes('id')) {
        //     return;
        //   } else {
        //     this.mapFields[this.primaryField] = this.mapFields['id'];
        //     delete this.mapFields['id'];
        //   }
        // }
        if (this.ngControl &&
            this.ngControl.formDirective &&
            this.ngControl.formDirective.form &&
            this.ngControl.formDirective.form.bindingData) {
            /** @type {?} */
            var bindingData = this.ngControl.formDirective.form.bindingData;
            if (bindingData.setValue) {
                /** @type {?} */
                var bindingPath = this.ngControl.formDirective.form.bindingPath;
                /** @type {?} */
                var pathArr = [];
                if (bindingPath) {
                    pathArr = bindingPath.split('/').filter((/**
                     * @param {?} n
                     * @return {?}
                     */
                    function (n) { return n !== ''; }));
                }
                /** @type {?} */
                var multiPath = this.mapFields['id'].split(this.multipleChoiceSeparator);
                /** @type {?} */
                var fieldPaths = pathArr.concat(multiPath[0].split('.'));
                /** @type {?} */
                var ids = bindingData.getValue(fieldPaths);
                if (!ids) {
                    this.emitSelections([]);
                    return;
                }
                this.assembleDataByOrgIds(ids.split(this.multipleChoiceSeparator));
            }
        }
    };
    // 根据传入的组织id数组获取对应的组织
    // 由ngOnChanges()、mapFields2Component()调用
    // 根据传入的组织id数组获取对应的组织
    // 由ngOnChanges()、mapFields2Component()调用
    /**
     * @private
     * @param {?} ids
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.assembleDataByOrgIds = 
    // 根据传入的组织id数组获取对应的组织
    // 由ngOnChanges()、mapFields2Component()调用
    /**
     * @private
     * @param {?} ids
     * @return {?}
     */
    function (ids) {
        var _this = this;
        if (!ids || !ids.length) {
            return;
        }
        if (ids && ids.length > 1 && !this.multiSelect) {
            this.againstSingleSelect = true;
            return;
        }
        this.apiService.getOrganizationsWithIds(ids).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            if (res && res.length) {
                _this.selections = res;
                if (!_this.mapFields) {
                    _this.assembleFinalText();
                    _this.emitSelections(_this.selections);
                    if (_this.dataPermission) {
                        _this.emitSelections(_this.selections);
                    }
                }
                else {
                    _this.emitSelections(_this.selections);
                }
            }
        }));
    };
    // 构造回显至input文本框的内容
    // 在有mapFields的前提下，由assembleDataByOrgIds()调用，clearAll()、removeItem()、returnResult()也会调用
    // 构造回显至input文本框的内容
    // 在有mapFields的前提下，由assembleDataByOrgIds()调用，clearAll()、removeItem()、returnResult()也会调用
    /**
     * @private
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.assembleFinalText = 
    // 构造回显至input文本框的内容
    // 在有mapFields的前提下，由assembleDataByOrgIds()调用，clearAll()、removeItem()、returnResult()也会调用
    /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.finalText = [];
        this.selections.forEach((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        function (item, index) {
            /** @type {?} */
            var formatedName = _this.formatFn && _this.formatFn(item);
            /** @type {?} */
            var finalText = '';
            if (_this.orgNameType == 'short') {
                finalText = formatedName ? formatedName : (item['abbreviation'] ? item['abbreviation'] : item[_this.displayField]);
            }
            else {
                finalText = formatedName ? formatedName : item[_this.displayField];
            }
            _this.finalText.push(finalText);
        }));
        this.finalText = this.finalText.slice();
    };
    // 初始化弹窗id及高度
    // 由ngOnInit()调用
    // 初始化弹窗id及高度
    // 由ngOnInit()调用
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.initDialogHeight = 
    // 初始化弹窗id及高度
    // 由ngOnInit()调用
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var uuid = CommonUtil.newGuid();
        this.dialogId += uuid;
        /** @type {?} */
        var windowInnerHeight = window.innerHeight;
        this.dialogHeight = windowInnerHeight - 60;
        if (this.dialogHeight > 680) {
            this.dialogHeight = 680;
        }
    };
    // 监听selections的变更，变更包含三种形式：valueChange、append、remove
    // 由ngOnInit()调用
    // 监听selections的变更，变更包含三种形式：valueChange、append、remove
    // 由ngOnInit()调用
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.listenSelectionsChange = 
    // 监听selections的变更，变更包含三种形式：valueChange、append、remove
    // 由ngOnInit()调用
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.selectionsSub = this.stateMgr.selectionsSubject.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data.changeType === ChangeType.ValueChanged) {
                _this.selections = _.cloneDeep(data.value);
            }
            else if (data.changeType === ChangeType.Remove) {
                _this.selections = _this.selections.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return item['id'] !== data.value['id']; }));
            }
            else if (data.changeType === ChangeType.Append) {
                CommonUtil.findIndex(data.value['id'], 'id', _this.selections) === -1 ? _this.selections.push(data.value) : null;
            }
            _this.onlyParentFinalText && _this.assembleParentFinal();
            _this.cdRef.detectChanges();
        }));
    };
    // 根据showTabIds判断是否启用收藏功能、群组功能
    // 由ngOnInit()调用
    // 根据showTabIds判断是否启用收藏功能、群组功能
    // 由ngOnInit()调用
    /**
     * @private
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.initTabs = 
    // 根据showTabIds判断是否启用收藏功能、群组功能
    // 由ngOnInit()调用
    /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.showTabIds.includes(this.activeTabId)) {
            this.activeTabId = this.showTabIds.split(',')[0];
        }
        if (this.dataPermission) {
            this.showTabIds = this.organizationTabId;
            this.activeTabId = this.organizationTabId;
            this.orgTabActive = true;
            return;
        }
        this.currentTabId = this.activeTabId;
        if (this.showTabIds.indexOf(this.favoritesTabId) !== -1) {
            this.enableFavorites = true;
        }
        else {
            this.enableFavorites = false;
        }
        if (this.showTabIds.indexOf(this.groupTabId) !== -1) {
            this.enableGroup = true;
        }
        else {
            this.enableGroup = false;
        }
        if (this.activeTabId == this.organizationTabId) {
            this.orgTabActive = true;
        }
        else if (this.activeTabId == this.groupTabId) {
            this.groupTabActive = true;
        }
        // 启用群组功能扩展函数
        /*
        if (this.enableGroup) {
          const ccwork = window.top.document.createElement("script")
          ccwork.src = this.groupExtendUrl + '?vs=' + Math.random();
          window.document.head.appendChild(ccwork);
        }
        */
    };
    // 判断群组名称是否合法
    // 由ngOnInit()调用
    // 判断群组名称是否合法
    // 由ngOnInit()调用
    /**
     * @private
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.initGroupSearch = 
    // 判断群组名称是否合法
    // 由ngOnInit()调用
    /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.enableGroup) {
            this.groupTitleStream.pipe(debounceTime(500), distinctUntilChanged()).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                if (!!data) {
                    _this.validGroupTitle = true;
                }
                else {
                    _this.validGroupTitle = false;
                }
            }));
        }
    };
    // 当浏览器窗口大小变更时，变更弹窗的大小
    // 由ngAfterViewInit()调用
    // 当浏览器窗口大小变更时，变更弹窗的大小
    // 由ngAfterViewInit()调用
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.listenWindowResize = 
    // 当浏览器窗口大小变更时，变更弹窗的大小
    // 由ngAfterViewInit()调用
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var dialogElement = document.querySelector('#' + this.dialogId);
        /** @type {?} */
        var targetElement = null;
        if (!!dialogElement) {
            targetElement = dialogElement.querySelector('.modal-dialog');
        }
        window.addEventListener('resize', (/**
         * @return {?}
         */
        function () {
            if (!!targetElement) {
                /** @type {?} */
                var height = window.innerHeight - 60;
                if (height > 680) {
                    height = 680;
                }
                _this.dialogHeight = height;
                /** @type {?} */
                var innerHeight_1 = height + 'px';
                /** @type {?} */
                var marginTop = '-' + height / 2 + 'px';
                _this.render.setStyle(targetElement, 'height', innerHeight_1);
                _this.render.setStyle(targetElement, 'margin-top', marginTop);
            }
        }));
    };
    // 初始化组织收藏数据
    // 由showFarrisDialog()调用，即打开选组织弹窗时发送请求
    // 初始化组织收藏数据
    // 由showFarrisDialog()调用，即打开选组织弹窗时发送请求
    /**
     * @private
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.initFavorites = 
    // 初始化组织收藏数据
    // 由showFarrisDialog()调用，即打开选组织弹窗时发送请求
    /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.favoritesTool.absOrgType = this.absOrgType;
        this.favoritesTool.busNum = this.busNum;
        this.favoritesTool.idField = 'id';
        this.favoritesTool.absRfQueryUrl = this.absRfQueryUrl;
        this.favoritesTool.absRfModifyUrl = this.absRfModifyUrl;
        this.favoritesTool.getFavoritesList().subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            /** @type {?} */
            var resData = result ? ((/** @type {?} */ (result))).orgData : [];
            _this.favoritesTool.favoritesList = resData;
            _this.favoritesTool.setFavoritesList(resData);
        }));
    };
    // 打开选组织弹窗
    // 打开选组织弹窗
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.showFarrisDialog = 
    // 打开选组织弹窗
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.isShow = true;
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.dialog.show();
                if (_this.showTabIds.indexOf(_this.favoritesTabId) !== -1) {
                    _this.initFavorites();
                }
                if (_this.againstSingleSelect) {
                    _this.notifyService.warning(_this.localeService.getValue("organizationSelector.errorMessage.multiSelect"));
                    return;
                }
                // if (this.mapFields && !Object.keys(this.mapFields).includes(this.primaryField)) {
                //   this.notifyService.warning(this.localeService.getValue("organizationSelector.errorMessage.mapFields"));
                //   return;
                // }
                _this.stateMgr.emitSelections(_this.selections, ChangeType.ValueChanged, 'form');
            }));
        }));
    };
    // input文本框内点击清除按钮时触发
    // input文本框内点击清除按钮时触发
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.clearAll = 
    // input文本框内点击清除按钮时触发
    /**
     * @return {?}
     */
    function () {
        this.selections = [];
        this.assembleFinalText();
        this.onModelChange(this.finalText.join());
        this.onModelTouched(this.finalText.join());
        this.emitSelections([]);
        this.inputClear.emit();
    };
    // input文本框内删除某一个已选组织时触发
    // input文本框内删除某一个已选组织时触发
    /**
     * @param {?} data
     * @param {?} index
     * @param {?} event
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.removeItem = 
    // input文本框内删除某一个已选组织时触发
    /**
     * @param {?} data
     * @param {?} index
     * @param {?} event
     * @return {?}
     */
    function (data, index, event) {
        if (this.selections.length > 0) {
            if (this.onlyParentFinalText) {
                /** @type {?} */
                var removeIndex = this.selections.findIndex((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return item['id'] == data['id']; }));
                /** @type {?} */
                var removeData_1 = this.selections.splice(removeIndex, 1);
                if (data.children.length) {
                    /** @type {?} */
                    var afterRemoveChild = this.selections.filter((/**
                     * @param {?} item
                     * @return {?}
                     */
                    function (item) { return !item.path.startsWith(removeData_1[0].path); }));
                    this.selections = afterRemoveChild.slice();
                }
                this.assembleFinalText();
                this.emitSelections(this.selections);
            }
            else {
                this.selections.splice(index, 1);
                this.assembleFinalText();
                this.emitSelections(data, ChangeType.Remove);
            }
        }
        else {
            this.finalText = this.finalText.filter((/**
             * @param {?} text
             * @return {?}
             */
            function (text) { return text !== data; }));
        }
        this.onModelChange(this.finalText.join());
        this.onModelTouched(this.finalText.join());
        this.tagRemoved.emit({
            deleteItem: data,
            deleteIndex: index,
            selections: this.selections
        });
        event.stopPropagation();
        event.cancelBubble = true;
    };
    // 清空最近已选组织
    // 清空最近已选组织
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.clearSeletions = 
    // 清空最近已选组织
    /**
     * @return {?}
     */
    function () {
        this.stateMgr.emitSelections([], ChangeType.ValueChanged, 'form');
    };
    // 点击搜索按钮或者敲击回车时触发
    // 点击搜索按钮或者敲击回车时触发
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.onSearch = 
    // 点击搜索按钮或者敲击回车时触发
    /**
     * @return {?}
     */
    function () {
        if (!!this.searchText) {
            this.hideSearchPanel = false;
            this.onSearchToken = CommonUtil.newGuid();
        }
    };
    // 清空组织搜索内容时触发
    // 清空组织搜索内容时触发
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.clearSearch = 
    // 清空组织搜索内容时触发
    /**
     * @return {?}
     */
    function () {
        this.hideSearchPanel = true;
        this.onSearchToken = false;
        if (this.currentTabId = this.organizationTabId) {
            this.libOrgTree.farrisTree.clearAll();
            this.libOrgTree.farrisTree.rowHeight = 28;
            if (this.multiSelect) {
                this.libOrgTree.checkOrgTreeNodes();
            }
            else {
                this.libOrgTree.selectOrgTreeNode();
            }
        }
    };
    // 监听搜索框内输入的文本
    // 监听搜索框内输入的文本
    /**
     * @param {?} e
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.listenSearchText = 
    // 监听搜索框内输入的文本
    /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (e.code === 'Backspace' || e.code === 'Delete') {
            if (!this.searchText) {
                this.clearSearch();
            }
        }
    };
    // 监听搜索页的关闭事件，目前只会接收到true
    // 监听搜索页的关闭事件，目前只会接收到true
    /**
     * @param {?} e
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.listenClosePanelChange = 
    // 监听搜索页的关闭事件，目前只会接收到true
    /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.hideSearchPanel = e;
        // 关闭搜索结果页后，需要更新树节点的是否选中状态
        if (this.currentTabId == this.organizationTabId) {
            this.libOrgTree.farrisTree.clearAll();
            this.libOrgTree.farrisTree.rowHeight = 28;
            if (this.multiSelect) {
                this.libOrgTree.checkOrgTreeNodes();
            }
            else {
                this.libOrgTree.selectOrgTreeNode();
            }
        }
    };
    // 关闭保存群组的弹窗时点击“取消”或者右上角叉号触发
    // 关闭保存群组的弹窗时点击“取消”或者右上角叉号触发
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.closeGroupDialog = 
    // 关闭保存群组的弹窗时点击“取消”或者右上角叉号触发
    /**
     * @return {?}
     */
    function () {
        this.groupDialogOpend = false;
        this.editingGroupId = null;
        this.groupTitle = null;
        this.groupNote = null;
        this.groupDialogTitle = this.localeService.getValue("organizationSelector.group.saveGroup");
    };
    // 在保存群组的弹窗内，修改群组名称时触发，监听keyup事件
    // 在保存群组的弹窗内，修改群组名称时触发，监听keyup事件
    /**
     * @param {?} e
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.listenGroupTitle = 
    // 在保存群组的弹窗内，修改群组名称时触发，监听keyup事件
    /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.groupTitleStream.next(this.groupTitle);
    };
    // 选组织的弹窗内有一个“已选”标志，鼠标移上去后会显示已选组织的页面，该页面内有“保存为群组”的按钮，点击后可以打开保存群组的弹窗
    // 选组织的弹窗内有一个“已选”标志，鼠标移上去后会显示已选组织的页面，该页面内有“保存为群组”的按钮，点击后可以打开保存群组的弹窗
    /**
     * @param {?} e
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.listenOpenGroup = 
    // 选组织的弹窗内有一个“已选”标志，鼠标移上去后会显示已选组织的页面，该页面内有“保存为群组”的按钮，点击后可以打开保存群组的弹窗
    /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.groupDialogOpend = e;
        this.groupDialogTitle = this.localeService.getValue("organizationSelector.group.saveGroup");
    };
    // 点击群组编辑按钮触发，可修改选中群组的名称及描述
    // 点击群组编辑按钮触发，可修改选中群组的名称及描述
    /**
     * @param {?} item
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.listenGroupChange = 
    // 点击群组编辑按钮触发，可修改选中群组的名称及描述
    /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.groupDialogOpend = true;
        this.editingGroupId = item.id;
        this.groupTitle = item.title;
        this.groupNote = item.description;
        this.groupDialogTitle = this.localeService.getValue("organizationSelector.group.editGroup");
    };
    // 关闭保存群组的弹窗时点击“确定”触发，将当前已选组织保存为一个新的群组
    // 关闭保存群组的弹窗时点击“确定”触发，将当前已选组织保存为一个新的群组
    /**
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.saveGroup = 
    // 关闭保存群组的弹窗时点击“确定”触发，将当前已选组织保存为一个新的群组
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.groupTitle) {
            this.validGroupTitle = false;
            return;
        }
        /** @type {?} */
        var groupId = CommonUtil.newGuid();
        /** @type {?} */
        var groupCode = 'orgGroup_private_';
        groupCode += Math.random().toString(16).substring(2, 10);
        /** @type {?} */
        var addIds = [];
        this.selections.forEach((/**
         * @param {?} selection
         * @return {?}
         */
        function (selection) {
            addIds.push(selection['id']);
        }));
        if (addIds.length > 1000) {
            this.messageService.show('warning', this.localeService.getValue("organizationSelector.errorMessage.limit"), true);
            this.groupDialogOpend = false;
            this.groupTitle = null;
            this.groupNote = null;
            return;
        }
        this.groupDialogOpend = false;
        /** @type {?} */
        var body = {
            groupId: groupId,
            groupCode: groupCode,
            title: this.groupTitle,
            description: this.groupNote,
            absOrgType: this.absOrgType,
            changeType: Number(1)
        };
        if (this.editingGroupId) {
            body.groupId = this.editingGroupId;
            body.changeType = Number(2);
            delete body.groupCode;
            this.updateNameOrNote(body);
            return;
        }
        /** @type {?} */
        var newGroup$ = this.apiService.saveGroup(body);
        newGroup$.pipe(switchMap((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            if (res) {
                _this.libOrgGroup.selectedGroupId = groupId;
                _this.groupTitle = null;
                _this.groupNote = null;
                /** @type {?} */
                var addOrgsInNewGroup$ = _this.apiService.saveGroup({ groupId: groupId, addIds: addIds, deleteIds: [], changeType: Number(3) });
                return addOrgsInNewGroup$.pipe(tap((/**
                 * @param {?} data
                 * @return {?}
                 */
                function (data) {
                    if (data) {
                        _this.libOrgGroup.getGroupList({ saveGroup: true });
                    }
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    throw new Error(_this.localeService.getValue("organizationSelector.errorMessage.saveOrganization"));
                })));
            }
        })), tap((/**
         * @return {?}
         */
        function () {
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.messageService.show('error', _this.localeService.getValue("organizationSelector.errorMessage.saveGroup"), true);
        }))).subscribe();
    };
    /**
     * @private
     * @param {?} param
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.updateNameOrNote = /**
     * @private
     * @param {?} param
     * @return {?}
     */
    function (param) {
        var _this = this;
        /** @type {?} */
        var updateInfo$ = this.apiService.saveGroup(param);
        updateInfo$.pipe(tap((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                _this.editingGroupId = null;
                _this.groupTitle = null;
                _this.groupNote = null;
                _this.groupDialogTitle = _this.localeService.getValue("organizationSelector.group.saveGroup");
                _this.libOrgGroup.getGroupList({ updateNameOrNote: true });
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.messageService.show('error', _this.localeService.getValue("organizationSelector.errorMessage.editGroup"), true);
        }))).subscribe();
    };
    // 点击选组织弹窗的确定或者取消按钮时触发，1为确定，0为取消
    // 点击选组织弹窗的确定或者取消按钮时触发，1为确定，0为取消
    /**
     * @param {?} type
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.returnResult = 
    // 点击选组织弹窗的确定或者取消按钮时触发，1为确定，0为取消
    /**
     * @param {?} type
     * @return {?}
     */
    function (type) {
        /** @type {?} */
        var finalText = '';
        if (!!type) {
            this.confirmToken = true;
            this.assembleFinalText();
            finalText = this.finalText.join();
            if (this.selections && this.selections.length) {
                this.selectionsChange.emit({
                    selections: this.selections
                });
                this.onModelChange(finalText);
                this.onModelTouched(finalText);
            }
            else {
                this.onModelChange(finalText);
                this.onModelTouched(finalText);
                this.selectionsChange.emit({
                    selections: this.selections
                });
            }
        }
        else {
            if (this.mapFields) {
                this.mapFields2Component();
            }
            else {
                this.assembleDataByOrgIds(this.orgIds);
            }
        }
        // this.orgIds = [];
        this.dialog.close();
        this.isShow = false;
    };
    /**
     * @param {?} e
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.changeTabs = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (!this.hideSearchPanel) {
            this.hideSearchPanel = true;
            this.onSearchToken = false;
        }
        this.currentTabId = e.nextId;
        if (this.showTabIds.includes(this.organizationTabId)) {
            this.libOrgTree.currentTabId = this.currentTabId;
        }
        if (e.nextId === this.groupTabId) {
            if (!this.libOrgGroup.groupData.public.length && !this.libOrgGroup.groupData.personal.length) {
                this.libOrgGroup.assembleData(true);
            }
        }
        else if (e.nextId === this.organizationTabId) {
            if (this.libOrgTree.initLoad) {
                this.libOrgTree.loadTreeData();
            }
            else {
                this.libOrgTree.farrisTree.clearAll();
                this.libOrgTree.farrisTree.rowHeight = 28;
                if (this.multiSelect) {
                    this.libOrgTree.checkOrgTreeNodes();
                }
                else {
                    this.libOrgTree.selectOrgTreeNode();
                }
            }
        }
    };
    /**
     * @param {?} e
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.quickShowDialog = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (!this.quickSelect) {
            return;
        }
        if (e.target.classList.contains('form-control')) {
            this.showFarrisDialog();
        }
    };
    /**
     * @private
     * @return {?}
     */
    AdminOrganizationSelectorComponent.prototype.assembleParentFinal = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.assembleFinalText();
        this.parentFinal = [];
        this.selections.forEach((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        function (item, index) {
            /** @type {?} */
            var highLayerOrg = _this.selections.filter((/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return item.path.startsWith(data.path) && item.path != data.path; }));
            /** @type {?} */
            var targetLayer;
            highLayerOrg.forEach((/**
             * @param {?} org
             * @return {?}
             */
            function (org) {
                if (!targetLayer || org['layer'] < targetLayer) {
                    targetLayer = org['layer'];
                }
            }));
            /** @type {?} */
            var highLayerOrgIndex = _this.selections.findIndex((/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return item.path.startsWith(data.path) && item.path != data.path && data.layer == targetLayer; }));
            if (highLayerOrgIndex > -1) {
                /** @type {?} */
                var parentIndex = _this.parentFinal.findIndex((/**
                 * @param {?} p
                 * @return {?}
                 */
                function (p) { return p.id == _this.selections[highLayerOrgIndex].id; }));
                if (parentIndex > -1) {
                    _this.parentFinal[parentIndex]['children'].push(_this.finalText[index]);
                }
                else {
                    _this.parentFinal.push(Object.assign({}, {
                        id: _this.selections[highLayerOrgIndex]['id'],
                        path: _this.selections[highLayerOrgIndex]['path'],
                        name: _this.finalText[highLayerOrgIndex],
                        children: [_this.finalText[index]]
                    }));
                }
            }
            else {
                /** @type {?} */
                var parentIndex = _this.parentFinal.findIndex((/**
                 * @param {?} p
                 * @return {?}
                 */
                function (p) { return p.id == item.id; }));
                if (parentIndex == -1) {
                    _this.parentFinal.push(Object.assign({}, {
                        id: item.id,
                        path: item.path,
                        name: _this.finalText[index],
                        children: []
                    }));
                }
            }
        }));
        this.abbreParentFinal = this.parentFinal.slice();
        this.abbreParentFinal.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (item.children.length > 10) {
                /** @type {?} */
                var firstChilds = item.children.slice(0, 5).concat(['……']);
                /** @type {?} */
                var lastChilds = item.children.slice(-5);
                item.children = firstChilds.concat(lastChilds || []);
            }
        }));
    };
    AdminOrganizationSelectorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-adminorganization-selector',
                    template: "<div #tagbox class=\"input-group s-input-group\" *ngIf=\"showInput\" [class.farris-input-group-readonly]=\"readonly\" [ngClass]=\"{'no-border': isTableTemplate}\">\r\n  <div class=\"form-control f-cmp-inputgroup--multi-wrapper multi-more\" [ngClass]=\"{'table-template-input': isTableTemplate}\"\r\n    (click)=\"quickShowDialog($event)\" *ngIf=\"!onlyParentFinalText\">\r\n    <div class=\"multi--content\" *ngFor=\"let item of finalText; index as i\" [title]=\"item\">\r\n      <span class=\"multi--item\">\r\n        {{item}}\r\n      </span>\r\n      <i class=\"f-icon f-icon-close cursor-pointer\" *ngIf=\"!readonly\" style=\"margin-top: 1px;\"\r\n        (click)=\"removeItem(item, i, $event)\"></i>\r\n    </div>\r\n    <span class=\"multi-another\" *ngIf=\"finalText.length === 0\">{{placeholder}}</span>\r\n  </div>\r\n  <div class=\"form-control f-cmp-inputgroup--multi-wrapper multi-more\" (click)=\"quickShowDialog($event)\" *ngIf=\"onlyParentFinalText\">\r\n    <div class=\"multi--content\" *ngFor=\"let item of abbreParentFinal; index as i\" [title]=\"item.name\">\r\n      <span class=\"multi--item\" farrisPopover [placement]=\"'bottom'\" [popover]=\"item.children.length ? item.children.join('\uFF0C') : '\u65E0'\"\r\n        [triggers]=\"'hover'\" [popoverTitle]=\"'\u5F53\u524D\u7EC4\u7EC7\u7684\u5DF2\u9009\u4E0B\u7EA7\u7EC4\u7EC7'\">\r\n        {{item.name}}\r\n      </span>\r\n      <i class=\"f-icon f-icon-close cursor-pointer\" *ngIf=\"!readonly\" style=\"margin-top: 1px;\"\r\n        (click)=\"removeItem(item, i, $event)\"></i>\r\n    </div>\r\n    <span class=\"multi-another\" *ngIf=\"finalText.length === 0\">{{placeholder}}</span>\r\n  </div>\r\n  <div *ngIf=\"!readonly\" class=\"input-group-append\" title=\"\">\r\n    <span *ngIf=\"finalText && finalText.length > 0\" class=\"input-group-text input-group-clear cursor-pointer\"\r\n      (click)=\"clearAll()\">\r\n      <i class=\"f-icon modal_close\"></i>\r\n    </span>\r\n    <span class=\"input-group-text cursor-pointer\" (click)=\"showFarrisDialog()\">\r\n      <span class=\"f-icon f-icon-lookup\"></span>\r\n    </span>\r\n  </div>\r\n</div>\r\n\r\n<farris-dialog #farrisDialog *ngIf=\"isShow\" class=\"fos-dialog\" [resizable]=\"false\" id=\"{{dialogId}}\"\r\n  [draggable]=\"draggable\" [buttons]=\"buttonRef\" [beforeClose]=\"farrisDialogbeforeClose\" [showMaxButton]=\"false\"\r\n  [width]=\"760\" [height]=\"dialogHeight\" [title]=\"'organizationSelector.selection.selectionDialogTitle' | locale\">\r\n  <div class=\"fos-content\">\r\n    <div class=\"fos-content-container\">\r\n      <div class=\"fos-content-main\">\r\n        <lib-search #searchResultPage [primaryField]=\"primaryField\" [multiSelect]=\"multiSelect\"\r\n          [enableFavorites]=\"enableFavorites\" [absOrgType]=\"absOrgType\" [notGetStopOrg]=\"notGetStopOrg\"\r\n          [onSearch]=\"onSearchToken\" [searchText]=\"searchText\" [pageSize]=\"searchPageSize\"\r\n          [onlyDisplayCompany]=\"onlyDisplayCompany\" [onlyCurrentAdminOrg]=\"onlyCurrentAdminOrg\"\r\n          (closePanelChange)=\"listenClosePanelChange($event)\" [hidden]=\"hideSearchPanel\">\r\n        </lib-search>\r\n        <input-group class=\"fos-search-input\" id=\"org-search-input\" [placeholder]=\"'organizationSelector.selection.searchPlaceHolder' | locale\"\r\n          [(value)]=\"searchText\" [groupText]=\"searchIcon\" (enterHandle)=\"onSearch()\" (clickHandle)=\"onSearch()\"\r\n          (keyup)=\"listenSearchText($event)\" (clear)=\"clearSearch()\" *ngIf=\"!dataPermission\">\r\n        </input-group>\r\n        <farris-tabs #selectTabs class=\"fos-tabs\" [value]=\"activeTabId\" [width]=\"'100%'\" [contentFill]=\"true\" (tabChange)=\"changeTabs($event)\">\r\n          <!-- \u7EC4\u7EC7\u673A\u6784  START -->\r\n          <farris-tab [title]=\"'organizationSelector.selection.organization' | locale\" [id]=\"organizationTabId\" *ngIf=\"showTabIds.indexOf(organizationTabId) !== -1\">\r\n            <ng-template #content>\r\n              <lib-org #libOrgTree [enableHierarchicalLoading]=\"enableHierarchicalLoading\" [notGetStopOrg]=\"notGetStopOrg\"\r\n                [enableFavorites]=\"enableFavorites\" [primaryField]=\"primaryField\" [cascadeCheck]=\"cascadeCheck\"\r\n                [cascadeDown]=\"cascadeDown\" [cascadeUp]=\"cascadeUp\" [keepSelect]=\"keepSelect\" [virtualizedLoad]=\"virtualizedLoad\"\r\n                [multiSelect]=\"multiSelect\" [absOrgType]=\"absOrgType\" [layer]=\"layer\" [orgIds]=\"orgIds\"\r\n                [selectedLimit]=\"selectedLimit\" [showCheckAll]=\"showCheckAll\" [activeTabId]=\"activeTabId\" [isActiveTab]=\"orgTabActive\"\r\n                [dataPermission]=\"dataPermission\" [dataPermissionIds]=\"dataPermissionIds\" [onlyDisplayCompany]=\"onlyDisplayCompany\"\r\n                [onlyCurrentAdminOrg]=\"onlyCurrentAdminOrg\" [orgNameType]=\"orgNameType\">\r\n              </lib-org>\r\n            </ng-template>\r\n          </farris-tab>\r\n          <!-- \u7EC4\u7EC7\u673A\u6784  END -->\r\n          <!-- \u6211\u7684\u6536\u85CF  START -->\r\n          <farris-tab [title]=\"'organizationSelector.selection.favorites' | locale\" [id]=\"favoritesTabId\" *ngIf=\"showTabIds.indexOf(favoritesTabId) !== -1\">\r\n            <ng-template #content>\r\n              <lib-favorites #libOrgFavorites [primaryField]=\"primaryField\" [multiSelect]=\"multiSelect\"></lib-favorites>\r\n            </ng-template>\r\n          </farris-tab>\r\n          <!-- \u6211\u7684\u6536\u85CF  END -->\r\n          <!-- \u5E38\u7528\u7FA4\u7EC4\u5C55\u793A\u533A  START -->\r\n          <farris-tab [title]=\"'organizationSelector.selection.group' | locale\" [id]=\"groupTabId\" *ngIf=\"showTabIds.indexOf(groupTabId) !== -1\">\r\n            <ng-template #content>\r\n              <lib-group #libOrgGroup \r\n                [primaryField]=\"primaryField\" [absOrgType]=\"absOrgType\" [multiSelect]=\"multiSelect\"\r\n                [enableFavorites]=\"enableFavorites\" [groupApiUrl]=\"groupApiUrl\" [notGetStopOrg]=\"notGetStopOrg\"\r\n                [enableHierarchicalLoading]=\"enableHierarchicalLoading\" [cascadeCheck]=\"cascadeCheck\"\r\n                [cascadeDown]=\"cascadeDown\" [cascadeUp]=\"cascadeUp\" [keepSelect]=\"keepSelect\" [virtualizedLoad]=\"virtualizedLoad\"\r\n                [layer]=\"layer\" [selectedLimit]=\"selectedLimit\" [showCheckAll]=\"showCheckAll\" [isActiveTab]=\"groupTabActive\"\r\n                (groupChange)=\"listenGroupChange($event)\">\r\n              </lib-group>\r\n            </ng-template>\r\n          </farris-tab>\r\n          <!-- \u5E38\u7528\u7FA4\u7EC4\u5C55\u793A\u533A  END -->\r\n        </farris-tabs>\r\n      </div>\r\n    </div>\r\n\r\n  </div>\r\n  <div class=\"s-personnel-group\" *ngIf=\"groupDialogOpend\">\r\n    <div class=\"s-personnel-group-title\">\r\n      <span class=\"f-icon f-sidebar-close\" (click)=\"closeGroupDialog()\"></span>{{groupDialogTitle}}\r\n    </div>\r\n    <div class=\"f-form-layout farris-form farris-form-controls-inline\" style=\"margin-top: 22px;\">\r\n      <div class=\"col-12 col-md-12 col-xl-12 col-el-12\">\r\n        <div class=\"farris-group-wrap\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\" style=\"width: 4.2rem\" for=\"month_4c7f0115_7796_42b4_bf41_beb7299dd84f_caum\">\r\n              <span class=\"farris-label-info text-danger\">*</span>\r\n              <span class=\"farris-label-text\">{{'organizationSelector.group.newGroupName' | locale}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n                <input class=\"form-control\" id=\"month_4c7f0115_7796_42b4_bf41_beb7299dd84f_caum\"\r\n                (keyup)=\"listenGroupTitle($event)\" placeholder=\"{{'organizationSelector.group.newGroupPlaceHolder' | locale}}\"\r\n                [ngStyle]=\"{'border-color': !validGroupTitle ? 'red' : '#D8DCE6'}\" [(ngModel)]=\"groupTitle\" />\r\n              <!--\r\n                <language-textbox id=\"month_4c7f0115_7796_42b4_bf41_beb7299dd84f_caum\" \r\n                [ngStyle]=\"{'border-color': !validGroupTitle ? 'red' : '#D8DCE6'}\"\r\n                [languages]=\"languagesArray\" [editable]=\"true\" [enableClear]=\"true\" [(ngModel)]=\"groupTitle\"\r\n                (keyup)=\"listenGroupTitle($event)\">\r\n              </language-textbox>\r\n              -->\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <div class=\"col-12 col-md-12 col-xl-12 col-el-12\" style=\"margin-top: 4px;\">\r\n        <div class=\"farris-group-wrap\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\" style=\"width: 4.2rem\"\r\n              for=\"billNote_7b43d269_2dd3_4777_8cfb_bd0c721575cf_hh9s\">\r\n              <span class=\"farris-label-text\">{{'organizationSelector.group.newGroupNote' | locale}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <textarea class=\"form-control\" style=\"height: 66px;\" [(ngModel)]=\"groupNote\"\r\n                id=\"billNote_7b43d269_2dd3_4777_8cfb_bd0c721575cf_hh9s\" placeholder=\"{{'organizationSelector.group.newGroupPlaceHolder' | locale}}\"></textarea>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n    <div class=\"s-personnel-group-footer\">\r\n      <input type=\"button\" class=\"btn btn-primary fr\" style=\"width: 100px;\" value=\"{{'organizationSelector.selection.confirm' | locale}}\" (click)=\"saveGroup()\" />\r\n      <input type=\"button\" class=\"btn btn-secondary fr mr-3\" value=\"{{'organizationSelector.selection.cancel' | locale}}\" (click)=\"closeGroupDialog()\" />\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n\r\n<ng-template #buttonRef>\r\n  <div class=\"fos-dialog-footer\">\r\n    <span class=\"fosd-footer-selection\" (mouseover)=\"selectionsShow=true\"\r\n      (mouseout)=\"selectionsShow=false\">{{'organizationSelector.selection.firstSelectionText' | locale}}\r\n      <a>{{selections.length}}</a>{{'organizationSelector.selection.lastSelectionText' | locale}}</span>\r\n    <span class=\"fosd-footer-clear\" (click)=\"clearSeletions()\">{{'organizationSelector.selection.clear' | locale}}</span>\r\n    <lib-selection #selectedComponent class=\"fosd-selection-content\" [primaryField]=\"primaryField\"\r\n      [enableGroup]=\"enableGroup\" (mouseover)=\"selectionsShow=true\" (mouseout)=\"selectionsShow=false\"\r\n      [hidden]=\"!selectionsShow\" (openedChange)=\"listenOpenGroup($event)\"></lib-selection>\r\n  </div>\r\n  <input type=\"button\" class=\"btn btn-secondary\" value=\"{{'organizationSelector.selection.cancel' | locale}}\" (click)=\"returnResult(0)\" />\r\n  <input type=\"button\" class=\"btn btn-primary\" style=\"width: 100px;\" value=\"{{'organizationSelector.selection.confirm' | locale}}\" (click)=\"returnResult(1)\" />\r\n</ng-template>",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return AdminOrganizationSelectorComponent; })),
                            multi: true
                        },
                        FavoritesTool, ApiService, StateMgrTool
                    ],
                    styles: [".fos-dialog .fos-content{width:100%;height:100%;min-width:760px;overflow:hidden}.fos-dialog ::ng-deep .modal-content{border-radius:10px}.fos-dialog ::ng-deep .modal-header{color:#2d2f33;border-bottom:none;padding:15px 0 15px 24px}.fos-dialog ::ng-deep .actions{padding-right:22px;font-size:20px}.fos-dialog ::ng-deep .modal-title-label{font-size:17px;font-weight:600}.fos-dialog ::ng-deep .modal-footer{padding:17px 24px;background:#fff;border-top:none}.fos-dialog .fos-content .fos-content-container{display:-webkit-box;display:flex;width:calc(100% - 48px);height:100%;margin:0 24px;border:1px solid #dee1ea;border-radius:3px}.fos-dialog .fos-content .fos-content-container .fos-content-main{-webkit-box-flex:0;flex:0 0 100%;position:relative;display:-webkit-box;display:flex}.fos-dialog .fos-content .fos-content-container .fos-content-main .fos-search-input{position:absolute;top:7px;right:14px;color:#959ab1;z-index:99;width:320px}.fos-dialog .fos-content .fos-content-container .fos-content-main .fos-tabs{min-width:575px}.fos-dialog ::ng-deep .fos-tabs .nav-link{line-height:18px;padding:11px 0!important}.fos-dialog ::ng-deep .fos-tabs .farris-tabs-header{background:#f9f9f9}.fos-dialog ::ng-deep .fos-tabs .nav-item{margin-right:30px}.fos-dialog ::ng-deep .fos-tabs .nav{padding-left:14px;font-size:16px;font-weight:600}.fos-dialog ::ng-deep .fos-tabs .nav-item .st-tab-text{color:#6f8cb4}.fos-dialog ::ng-deep .fos-tabs .nav-item .active .st-tab-text{color:#2a87ff}.fos-dialog::ng-deep input::-webkit-input-placeholder{color:#b4bccc;font-size:13px;line-height:24px}.fos-dialog ::ng-deep input::-moz-input-placeholder{color:#b4bccc;font-size:13px;line-height:24px}.fos-dialog ::ng-deep input::-ms-input-placeholder{color:#b4bccc;font-size:13px;line-height:24px}.fos-dialog ::ng-deep .fos-search-input .form-control{line-height:26px;height:26px;font-size:13px}.fos-dialog ::ng-deep .fos-treetable .farris-treetable-thead{background:#fff;font-size:13px}.fos-dialog ::ng-deep .fos-treetable .farris-treetable-thead th .tt-checkbox .custom-control-label::before{top:2px;margin:0}.fos-dialog ::ng-deep .fos-treetable .farris-treetable-thead th:hover{background:#fff}.fos-dialog ::ng-deep .fos-treetable .treetable-cell-content span{font-size:13px}:host.f-organization-selector-tableTempl{display:-webkit-box;display:flex;flex-wrap:wrap;-webkit-box-align:center;align-items:center;height:100%}:host .input-group .input-group-append{background:#fff}:host .s-input-group .input-group-text{color:rgba(0,0,0,.25)}:host .cursor-pointer{cursor:pointer}:host .s-input-group .multi-more{padding:0 0 2px 8px;height:auto;min-height:24px;-webkit-box-flex:1;flex:1;overflow:hidden}:host .s-input-group .table-template-input{display:-webkit-box;display:flex;flex-wrap:wrap;-webkit-box-align:center;align-items:center;min-height:43px}:host .no-border{border:none}:host .s-input-group .multi--content{display:-webkit-box;display:flex;flex-wrap:nowrap;cursor:default;padding:0 5px;background:#ebf7fe;border:1px solid #cfedff;margin:3px 3px 0 0;float:left;height:17px;border-radius:3px;line-height:17px;-webkit-box-pack:justify;justify-content:space-between;box-sizing:content-box;max-width:calc(100% - 10px)}:host .s-input-group .multi--item{line-height:16px;text-overflow:ellipsis;height:16px;-webkit-box-flex:1;flex:1;overflow:hidden;word-break:break-all}:host .s-input-group .multi-another{padding:3px 3px 0 0;line-height:16px;float:left;height:17px;box-sizing:content-box;border:1px solid #fff;color:#bfbfbf}.fos-dialog ::ng-deep .fos-dialog-footer{width:200px;height:30px;line-height:30px;position:absolute;left:24px;text-align:left;cursor:default}.fos-dialog ::ng-deep .fos-dialog-footer .fosd-footer-selection{font-size:14px;color:#2d2f33;background:#f4f9ff;border-radius:5px;padding:2px 5px}.fos-dialog ::ng-deep .fos-dialog-footer .fosd-footer-selection a{font-size:17px;color:#2a87ff}.fos-dialog ::ng-deep .fos-dialog-footer .fosd-footer-clear{font-size:13px;color:#2a87ff;margin-left:1rem;cursor:pointer}.fos-dialog ::ng-deep .fos-dialog-footer .fosd-selection-content{position:absolute;bottom:25px;left:-10px;background:#fff;z-index:102}.fos-dialog::ng-deep .pagination-container{height:1.5rem;margin-bottom:21px}.fos-dialog::ng-deep .pagination-container input::-webkit-inner-spin-button,.fos-dialog::ng-deep .pagination-container input::-webkit-outer-spin-button{-webkit-appearance:none}.fos-dialog::ng-deep .pagination-container input[type=number]{-moz-appearance:textfield;text-align:center;width:38px;line-height:20px;border:1px solid #e9ecf3;border-radius:6px;box-shadow:0 2px 6px 0 rgba(31,35,41,.06);margin-left:4px;outline:0}.fos-dialog::ng-deep .pagination-container input[type=number]:focus{border:1px solid #0084ff}.fos-dialog::ng-deep .pagination-container li{line-height:22px}.fos-dialog ::ng-deep .fos-treetable td{border-color:#fff}.fos-dialog ::ng-deep .default-light-color{color:#f7b500}.fos-dialog ::ng-deep .default-dark-color{color:#d8dce6}.fos-dialog ::ng-deep .s-personnel-group{width:440px;height:250px;border-radius:5px;background:#fff;box-shadow:0 0 40px 0 rgba(0,0,0,.2);padding:16px 24px;position:fixed;top:50%;left:50%;margin-left:-220px;margin-top:-125px;z-index:9999}.fos-dialog ::ng-deep .s-personnel-group .s-personnel-group-title{font-size:17px;line-height:24px;color:#2d2f33;font-weight:600}.fos-dialog ::ng-deep .s-personnel-group .s-personnel-group-title .f-icon{float:right;font-size:20px;color:#878d99;margin-top:3px;cursor:pointer}.fos-dialog ::ng-deep .s-personnel-group .s-personnel-group-footer{margin-top:20px}.fos-dialog ::ng-deep .s-personnel-group .col-12{padding:0}.fos-dialog ::ng-deep .s-personnel-group-footer .title-checkall{margin-top:6px}.fos-dialog ::ng-deep .s-personnel-group .farris-label-text{color:#5a5e66}.fos-dialog ::ng-deep .s-personnel-group .s-personnel-group-footer .fr{float:right;height:32px}"]
                }] }
    ];
    /** @nocollapse */
    AdminOrganizationSelectorComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: Injector },
        { type: FavoritesTool },
        { type: ApiService },
        { type: NgZone },
        { type: StateMgrTool }
    ]; };
    AdminOrganizationSelectorComponent.propDecorators = {
        baseCls: [{ type: HostBinding, args: ['class.f-organization-selector',] }],
        bableTemplClass: [{ type: HostBinding, args: ['class.f-organization-selector-tableTempl',] }],
        width: [{ type: HostBinding, args: ['style.width',] }, { type: Input }],
        dialog: [{ type: ViewChild, args: ['farrisDialog',] }],
        libOrgGroup: [{ type: ViewChild, args: ['libOrgGroup',] }],
        libOrgTree: [{ type: ViewChild, args: ['libOrgTree',] }],
        libOrgFavorites: [{ type: ViewChild, args: ['libOrgFavorites',] }],
        isTableTemplate: [{ type: Input }],
        showInput: [{ type: Input }],
        busNum: [{ type: Input }],
        multiSelect: [{ type: Input }],
        placeholder: [{ type: Input }],
        searchPlaceholder: [{ type: Input }],
        readonly: [{ type: Input }],
        absUserOrgUrl: [{ type: Input }],
        absRfQueryUrl: [{ type: Input }],
        absRfModifyUrl: [{ type: Input }],
        groupApiUrl: [{ type: Input }],
        groupExtendUrl: [{ type: Input }],
        absOrgType: [{ type: Input }],
        enableHierarchicalLoading: [{ type: Input }],
        notGetStopOrg: [{ type: Input }],
        primaryField: [{ type: Input }],
        displayField: [{ type: Input }],
        orgIds: [{ type: Input }],
        layer: [{ type: Input }],
        cascadeCheck: [{ type: Input }],
        cascadeDown: [{ type: Input }],
        cascadeUp: [{ type: Input }],
        showCheckAll: [{ type: Input }],
        keepSelect: [{ type: Input }],
        formatFn: [{ type: Input }],
        selectedLimit: [{ type: Input }],
        showTabIds: [{ type: Input }],
        activeTabId: [{ type: Input }],
        searchPageSize: [{ type: Input }],
        mapFields: [{ type: Input }],
        multipleChoiceSeparator: [{ type: Input }],
        draggable: [{ type: Input }],
        virtualizedLoad: [{ type: Input }],
        dataPermission: [{ type: Input }],
        dataPermissionIds: [{ type: Input }],
        orgNameType: [{ type: Input }],
        onlyDisplayCompany: [{ type: Input }],
        onlyCurrentAdminOrg: [{ type: Input }],
        onlyParentFinalText: [{ type: Input }],
        quickSelect: [{ type: Input }],
        selectionsChange: [{ type: Output }],
        inputClear: [{ type: Output }],
        tagRemoved: [{ type: Output }]
    };
    return AdminOrganizationSelectorComponent;
}());
export { AdminOrganizationSelectorComponent };
if (false) {
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.baseCls;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.width;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.dialog;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.libOrgGroup;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.libOrgTree;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.libOrgFavorites;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.isTableTemplate;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.showInput;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.busNum;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.multiSelect;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.placeholder;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.searchPlaceholder;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.readonly;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.absUserOrgUrl;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.absRfQueryUrl;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.absRfModifyUrl;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.groupApiUrl;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.groupExtendUrl;
    /**
     * 数据源类型
     * 默认为： System_organization  系统数据源   获取系统人员、系统组织
     * Administrative_organization  已扩展数据源  获取行政组织、行政人员
     * absOrgType {string}
     * PersonnelSelectorComponent
     * @type {?}
     */
    AdminOrganizationSelectorComponent.prototype.absOrgType;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.enableHierarchicalLoading;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.notGetStopOrg;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.primaryField;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.displayField;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.orgIds;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.layer;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.cascadeCheck;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.cascadeDown;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.cascadeUp;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.showCheckAll;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.keepSelect;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.formatFn;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.selectedLimit;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.showTabIds;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.activeTabId;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.searchPageSize;
    /**
     * 字段映射, 仅在farris form中使用{ 帮助数据字段：表单字段 }
     * @type {?}
     */
    AdminOrganizationSelectorComponent.prototype.mapFields;
    /**
     * 多选分隔符
     * @type {?}
     */
    AdminOrganizationSelectorComponent.prototype.multipleChoiceSeparator;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.draggable;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.virtualizedLoad;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.dataPermission;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.dataPermissionIds;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.orgNameType;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.onlyDisplayCompany;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.onlyParentFinalText;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.quickSelect;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.selectionsChange;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.inputClear;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.tagRemoved;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.selections;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.originalSelections;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.finalText;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.parentFinal;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.abbreParentFinal;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.lookupIcon;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.searchIcon;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.organizationTabId;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.groupTabId;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.favoritesTabId;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.dialogHeight;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.dialogId;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.isShow;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.selectionsSub;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.confirmToken;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.hideSearchPanel;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.searchText;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.onSearchToken;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.enableFavorites;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.enableGroup;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.groupDialogOpend;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.groupDialogTitle;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.groupTitle;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.groupNote;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.validGroupTitle;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.groupTitleStream;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.ngControl;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.selectionsShow;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.editingGroupId;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.cdRef;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.currentTabId;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.notifyService;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.orgTabActive;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.groupTabActive;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.againstSingleSelect;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.localeService;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.languagesArray;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype._onlyCurrentAdminOrg;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.farrisDialogbeforeClose;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.onModelChange;
    /** @type {?} */
    AdminOrganizationSelectorComponent.prototype.onModelTouched;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.render;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.favoritesTool;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.apiService;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    AdminOrganizationSelectorComponent.prototype.stateMgr;
}
