// @gsp-cmp/ubcc-param-data-ui/esm2015/lib/controls/dimension-control/dimension-input-control/dimension-input-control.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/controls/dimension-control/dimension-input-control/dimension-input-control.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { UbccEventBusService } from '../../../service/event-bus-service';
export class DimensionInputControlComponent {
    /**
     * @param {?} ebsService
     */
    constructor(ebsService) {
        this.ebsService = ebsService;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.dimensionItem = this.dimensionItem || {};
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    ngModelChangeHandler($event) {
        this.ebsService.ebs.emit(this.dimensionItem);
    }
}
DimensionInputControlComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-dimension-input-control',
                template: "<div class=\"farris-form-controls-inline farris-group-wrap col-12\">\n  <div class=\"form-group farris-form-group\">\n    <label class=\"col-form-label col-4\">\n      <span class=\"farris-label-text\" style=\"float:right\">{{dimensionItem.name}}</span>\n    </label>\n    <div class=\"farris-input-wrap col-6\">\n      <input type=\"text\" class=\"form-control\" placeholder=\"\u8BF7\u8F93\u5165\" autocomplete=\"off\" (ngModelChange)=\"ngModelChangeHandler($event)\"\n      [(ngModel)]=\"dimensionItem.value\" name=\"dimensionItem{{dimensionItem.code}}\">\n    </div>\n  </div>\n</div>",
                providers: [UbccEventBusService],
                styles: [""]
            }] }
];
/** @nocollapse */
DimensionInputControlComponent.ctorParameters = () => [
    { type: UbccEventBusService }
];
DimensionInputControlComponent.propDecorators = {
    dimensionItem: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DimensionInputControlComponent.prototype.dimensionItem;
    /**
     * @type {?}
     * @private
     */
    DimensionInputControlComponent.prototype.ebsService;
}
