// @farris/ui-datagrid-filter/esm5/lib/datagrid-filter-cell-editor.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FormGroup } from '@angular/forms';
import { Directive, Input, Injector, ComponentFactoryResolver, ViewContainerRef, Renderer2 } from '@angular/core';
import { ColumnFilterType } from '@farris/ui-datagrid';
import { FilterTextboxComponent } from './filter-editors/filter-row-textbox.component';
import { FilterCheckboxComponent } from './filter-editors/filter-row-checkbox.component';
var FilterCellEditorDirective = /** @class */ (function () {
    function FilterCellEditorDirective(injector, resolver, container, render) {
        this.injector = injector;
        this.resolver = resolver;
        this.container = container;
        this.render = render;
        this.delay = 300;
    }
    /**
     * @return {?}
     */
    FilterCellEditorDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.createFilterEditor();
    };
    /**
     * @private
     * @return {?}
     */
    FilterCellEditorDirective.prototype.createFilterEditor = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.column.filter) {
            /** @type {?} */
            var columnFilter = (/** @type {?} */ (this.column.filter));
            columnFilter.type = columnFilter.type || ColumnFilterType.string;
            /** @type {?} */
            var constructorCmp = FilterTextboxComponent;
            switch (columnFilter.type) {
                case ColumnFilterType.boolean:
                    constructorCmp = FilterCheckboxComponent;
                    break;
                // case ColumnFilterType.enum:
                //     constructorCmp = FilterEnumComponent;
                //     break;
            }
            /** @type {?} */
            var factory = this.resolver.resolveComponentFactory(constructorCmp);
            this.componentRef = this.container.createComponent(factory);
            this.componentRef.instance.column = this.column;
            this.componentRef.instance.group = this.group;
            this.componentRef.instance.gridInstance = this.gridInstance;
            this.componentRef.instance.disabled = columnFilter.type >= 3;
            this.componentRef.instance.delay = this.delay;
            if (this.gridInstance.conditions && Object.keys(this.gridInstance.conditions).length) {
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    _this.componentRef.instance.initFilter(_this.gridInstance.conditions[_this.column.field]);
                }), 100);
            }
        }
    };
    /**
     * @return {?}
     */
    FilterCellEditorDirective.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        if (this.componentRef) {
            this.componentRef.instance.clearValue();
        }
    };
    FilterCellEditorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[filter-cell-editor]',
                },] }
    ];
    /** @nocollapse */
    FilterCellEditorDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: ViewContainerRef },
        { type: Renderer2 }
    ]; };
    FilterCellEditorDirective.propDecorators = {
        column: [{ type: Input }],
        group: [{ type: Input }],
        gridInstance: [{ type: Input }],
        delay: [{ type: Input }]
    };
    return FilterCellEditorDirective;
}());
export { FilterCellEditorDirective };
if (false) {
    /** @type {?} */
    FilterCellEditorDirective.prototype.column;
    /** @type {?} */
    FilterCellEditorDirective.prototype.group;
    /** @type {?} */
    FilterCellEditorDirective.prototype.gridInstance;
    /** @type {?} */
    FilterCellEditorDirective.prototype.delay;
    /** @type {?} */
    FilterCellEditorDirective.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    FilterCellEditorDirective.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    FilterCellEditorDirective.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    FilterCellEditorDirective.prototype.container;
    /**
     * @type {?}
     * @private
     */
    FilterCellEditorDirective.prototype.render;
}
