// @farris/ui-datagrid-filter/esm5/lib/datagrid-filter-row.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { TimePickerModule } from '@farris/ui-time-picker';
import { LocaleModule } from '@farris/ui-locale';
import { DatagridModule, GRID_FILTER_ROW } from '@farris/ui-datagrid';
import { FarrisDatePickerModule } from '@farris/ui-datepicker';
import { NumberSpinnerModule } from '@farris/ui-number-spinner';
import { DatalistModule } from '@farris/ui-datalist';
import { FarrisFormsModule } from '@farris/ui-forms';
import { FarrisCommonModule } from '@farris/ui-common';
import { FilterRowComponent } from './datagrid-filter-row.component';
import { FilterCellEditorDirective } from './datagrid-filter-cell-editor.directive';
import { FilterTextboxComponent } from './filter-editors/filter-row-textbox.component';
import { FilterRowPanelComponent } from './filter-editors/filter-row-panel.component';
import { FilterCheckboxComponent } from './filter-editors/filter-row-checkbox.component';
import { ConditionEditorComponent } from './filter-editors/condition-editor.component';
import { FilterDatalistComponent } from './filter-editors/filter-datalist.component';
import { FilterRowEditorComponent } from './filter-editors/filter-row-editor.component';
import { FilterSelectComponent } from './filter-editors/filter-row-select.component';
/** @type {?} */
export var FilterRowProvider = { provide: GRID_FILTER_ROW, useValue: FilterRowComponent };
var DatagridFilterRowModule = /** @class */ (function () {
    function DatagridFilterRowModule() {
    }
    /**
     * @return {?}
     */
    DatagridFilterRowModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: DatagridFilterRowModule,
            providers: [
                { provide: GRID_FILTER_ROW, useValue: FilterRowComponent }
            ]
        };
    };
    DatagridFilterRowModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        FilterRowComponent,
                        FilterCellEditorDirective,
                        FilterTextboxComponent,
                        FilterCheckboxComponent,
                        FilterRowPanelComponent,
                        FilterDatalistComponent,
                        FilterRowEditorComponent,
                        FilterSelectComponent,
                        ConditionEditorComponent
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        DatagridModule,
                        FarrisDatePickerModule,
                        NumberSpinnerModule,
                        DatalistModule,
                        TimePickerModule,
                        LocaleModule.forRoot(),
                        FarrisFormsModule,
                        FarrisCommonModule
                    ],
                    entryComponents: [
                        FilterRowComponent,
                        FilterTextboxComponent,
                        FilterCheckboxComponent,
                        ConditionEditorComponent,
                        FilterDatalistComponent,
                        FilterRowPanelComponent,
                        FilterRowEditorComponent,
                        FilterSelectComponent
                    ],
                    exports: [
                        FilterRowComponent
                    ]
                },] }
    ];
    return DatagridFilterRowModule;
}());
export { DatagridFilterRowModule };
