// @gsp-aif/common-api/fesm5/gsp-aif-common-api.js
import { __extends } from 'tslib';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var AifCallType = {
    /**
     * 默认值
     */
    Default: 0,
    /**
     * 同步调用
     */
    Sync: 1,
    /**
     * 异步调用
     */
    Async: 2,
};
AifCallType[AifCallType.Default] = 'Default';
AifCallType[AifCallType.Sync] = 'Sync';
AifCallType[AifCallType.Async] = 'Async';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var AifBizActionType = {
    /**
     * 生单
     */
    Creation: 1,
    /**
     * 回写
     */
    Writeback: 2,
};
AifBizActionType[AifBizActionType.Creation] = 'Creation';
AifBizActionType[AifBizActionType.Writeback] = 'Writeback';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 实体抽象类
 * @abstract
 */
var  /**
 * 实体抽象类
 * @abstract
 */
EntityBase = /** @class */ (function () {
    function EntityBase() {
    }
    return EntityBase;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 元数据实体抽象类
 * @abstract
 */
var  /**
 * 元数据实体抽象类
 * @abstract
 */
MetadataEntityBase = /** @class */ (function (_super) {
    __extends(MetadataEntityBase, _super);
    function MetadataEntityBase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return MetadataEntityBase;
}(EntityBase));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 业务动作实体类
 */
var  /**
 * 业务动作实体类
 */
AifBizActionEntity = /** @class */ (function (_super) {
    __extends(AifBizActionEntity, _super);
    function AifBizActionEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return AifBizActionEntity;
}(MetadataEntityBase));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 业务场景图实体类
 */
var  /**
 * 业务场景图实体类
 */
AifBizFlowchartEntity = /** @class */ (function (_super) {
    __extends(AifBizFlowchartEntity, _super);
    function AifBizFlowchartEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return AifBizFlowchartEntity;
}(MetadataEntityBase));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 业务事件实体类
 */
var  /**
 * 业务事件实体类
 */
AifBizEventEntity = /** @class */ (function (_super) {
    __extends(AifBizEventEntity, _super);
    function AifBizEventEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return AifBizEventEntity;
}(MetadataEntityBase));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 点到点元数据实体抽象类
 * @abstract
 */
var  /**
 * 点到点元数据实体抽象类
 * @abstract
 */
P2PMetadataEntityBase = /** @class */ (function (_super) {
    __extends(P2PMetadataEntityBase, _super);
    function P2PMetadataEntityBase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return P2PMetadataEntityBase;
}(MetadataEntityBase));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 映射规则实体抽象类
 * @abstract
 */
var  /**
 * 映射规则实体抽象类
 * @abstract
 */
MappingEntity = /** @class */ (function (_super) {
    __extends(MappingEntity, _super);
    function MappingEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return MappingEntity;
}(P2PMetadataEntityBase));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 生单规则实体类
 */
var  /**
 * 生单规则实体类
 */
AifCreationRuleEntity = /** @class */ (function (_super) {
    __extends(AifCreationRuleEntity, _super);
    function AifCreationRuleEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return AifCreationRuleEntity;
}(MappingEntity));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 回写规则实体类
 */
var  /**
 * 回写规则实体类
 */
AifWritebackRuleEntity = /** @class */ (function (_super) {
    __extends(AifWritebackRuleEntity, _super);
    function AifWritebackRuleEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return AifWritebackRuleEntity;
}(MappingEntity));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 业务流程实体抽象类
 * @abstract
 */
var  /**
 * 业务流程实体抽象类
 * @abstract
 */
BizProcessEntity = /** @class */ (function (_super) {
    __extends(BizProcessEntity, _super);
    function BizProcessEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return BizProcessEntity;
}(P2PMetadataEntityBase));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 生单流程实体类
 */
var  /**
 * 生单流程实体类
 */
AifCreationProcessEntity = /** @class */ (function (_super) {
    __extends(AifCreationProcessEntity, _super);
    function AifCreationProcessEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return AifCreationProcessEntity;
}(BizProcessEntity));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 回写流程实体类
 */
var  /**
 * 回写流程实体类
 */
AifWritebackProcessEntity = /** @class */ (function (_super) {
    __extends(AifWritebackProcessEntity, _super);
    function AifWritebackProcessEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return AifWritebackProcessEntity;
}(BizProcessEntity));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { AifCallType, AifBizActionType, AifBizActionEntity, AifBizFlowchartEntity, AifBizEventEntity, AifCreationRuleEntity, AifWritebackRuleEntity, AifCreationProcessEntity, AifWritebackProcessEntity, BizProcessEntity as ɵe, EntityBase as ɵb, MappingEntity as ɵc, MetadataEntityBase as ɵa, P2PMetadataEntityBase as ɵd };
