// @farris/external-integration/esm5/lib/directives/ext-integration-data-mapping.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ext-integration-data-mapping.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, Input, Optional, Output } from "@angular/core";
import { ViewModel } from "@farris/devkit";
import { ExternalIntegrationComponent } from "../../lib/component/external-integration/external-integration.component";
var ExtIntegrationDataMappingDirective = /** @class */ (function () {
    function ExtIntegrationDataMappingDirective(vm, extIntegrationComponent) {
        this.vm = vm;
        this.extIntegrationComponent = extIntegrationComponent;
        this.showModal = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ExtIntegrationDataMappingDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.extIntegrationComponent.selectSapData.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.mappingData(data, _this.myField);
        }));
    };
    /**
     * 字段映射
     * @param sapData sap返回信息
     * @param mappingField 映射关系
     */
    /**
     * 字段映射
     * @param {?} sapData sap返回信息
     * @param {?} mappingField 映射关系
     * @return {?}
     */
    ExtIntegrationDataMappingDirective.prototype.mappingData = /**
     * 字段映射
     * @param {?} sapData sap返回信息
     * @param {?} mappingField 映射关系
     * @return {?}
     */
    function (sapData, mappingField) {
        var _this = this;
        /** @type {?} */
        var helpFields = Object.keys(mappingField);
        helpFields.forEach((/**
         * @param {?} fieldName
         * @return {?}
         */
        function (fieldName) {
            /** @type {?} */
            var pathArr = _this.getBindingPathArray();
            // 获取字段值
            /** @type {?} */
            var val;
            if (sapData) {
                val = sapData[fieldName];
            }
            _this.vm.bindingData.setValue(pathArr.concat(mappingField[fieldName]), val, true, true);
        }));
    };
    /**
     * @private
     * @return {?}
     */
    ExtIntegrationDataMappingDirective.prototype.getBindingPathArray = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var path = this.vm.bindingPath;
        if (path) {
            return path.split("/").filter((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return n !== ""; }));
        }
        return [];
    };
    ExtIntegrationDataMappingDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[extIntegration-dataMapping]",
                },] }
    ];
    /** @nocollapse */
    ExtIntegrationDataMappingDirective.ctorParameters = function () { return [
        { type: ViewModel, decorators: [{ type: Optional }] },
        { type: ExternalIntegrationComponent }
    ]; };
    ExtIntegrationDataMappingDirective.propDecorators = {
        myField: [{ type: Input, args: ["mapFields",] }],
        showModal: [{ type: Output }]
    };
    return ExtIntegrationDataMappingDirective;
}());
export { ExtIntegrationDataMappingDirective };
if (false) {
    /** @type {?} */
    ExtIntegrationDataMappingDirective.prototype.myField;
    /** @type {?} */
    ExtIntegrationDataMappingDirective.prototype.showModal;
    /**
     * @type {?}
     * @private
     */
    ExtIntegrationDataMappingDirective.prototype.vm;
    /**
     * @type {?}
     * @private
     */
    ExtIntegrationDataMappingDirective.prototype.extIntegrationComponent;
}
