// @farris/ide-devkit/esm2015/lib/ideframework/gsp/gsp-event-bus.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventBus } from '../eventbus';
export class GspEventBus extends EventBus {
    constructor() {
        super(...arguments);
        this.commands = [];
    }
    /**
     * @param {?} target
     * @param {?} tokenValue
     * @param {?} eventName
     * @param {?} caller
     * @param {?} handler
     * @return {?}
     */
    on(target, tokenValue, eventName, caller, handler) {
        // 如果有对应激活命令，监听事件时立即执行回调函数
        /** @type {?} */
        const command = this.commands.find(item => item.name === eventName);
        if (command) {
            handler.call(caller, command.eventeParams);
        }
        return super.on(target, tokenValue, eventName, caller, handler);
    }
    /**
     * @param {?} emitterType
     * @param {?} tokenValue
     * @param {?} eventName
     * @param {?} eventArgs
     * @return {?}
     */
    post(emitterType, tokenValue, eventName, eventArgs) {
        this.dispatchCommand(eventName, eventArgs);
        super.post(emitterType, tokenValue, eventName, eventArgs);
    }
    /**
     * @param {?} commandName
     * @param {?} plugin
     * @param {?} activateFn
     * @return {?}
     */
    registryActivateCommand(commandName, plugin, activateFn) {
        this.commands.push({
            name: commandName,
            plugin,
            activateFn
        });
    }
    /**
     * @private
     * @param {?} commandName
     * @param {?} args
     * @return {?}
     */
    dispatchCommand(commandName, args) {
        /** @type {?} */
        const command = this.commands.find(item => item.name === commandName);
        if (command) {
            command.activateFn(command.plugin);
            command.eventParams = args;
        }
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    GspEventBus.prototype.commands;
}
