// @farris/ide-devkit/esm2015/lib/ideframework/gsp/gsp.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CacheService } from './../../services/cache/services/cache.service';
import { CommandRegistry } from './command-registry';
import { Util } from '../util/util';
import { ViewRegistry } from './view-registry';
import { Workspace } from './workspace';
import { GspEventBus } from './gsp-event-bus';
import { Ide } from './ide/ide';
export class GSP {
    /**
     * @return {?}
     */
    get eventBus() { return this.pEventBus; }
    /**
     * @return {?}
     */
    get util() { return this.pUtil; }
    /**
     * @return {?}
     */
    get workspace() { return this.pWorkspace; }
    /**
     * @return {?}
     */
    get views() { return this.pViews; }
    /**
     * @return {?}
     */
    get commands() { return this.pCommands; }
    /**
     * @return {?}
     */
    get ide() { return this.pIde; }
    /**
     * @return {?}
     */
    get cache() { return this.pCache; }
    /**
     * @return {?}
     */
    get context() { return this.pContext; }
    /**
     * @param {?=} parent
     */
    constructor(parent) {
        this.pConfig = {
            /**
             * @param {?} key
             * @return {?}
             */
            get(key) {
                return this[key];
            }
        };
        /** @type {?} */
        let ide;
        if (parent) {
            this.pViews = parent.views;
            this.pCommands = parent.commands;
            this.pEventBus = parent.eventBus;
            this.pUtil = parent.util;
            this.pWorkspace = parent.workspace;
            this.pCache = parent.cache;
            ide = parent.ide;
            /** @type {?} */
            const ctor = parent.context.constructor;
            this.pContext = new ctor();
        }
        else {
            this.pViews = new ViewRegistry(this);
            this.pCommands = new CommandRegistry();
            this.pEventBus = new GspEventBus();
            this.pUtil = new Util();
            this.pWorkspace = new Workspace({
                config: this.pConfig,
                viewRegistry: this.pViews
            });
            this.pCache = new CacheService(null);
        }
        this.pIde = new Ide(ide);
    }
    /**
     * @deprecated 由子模块独自创建gsp实例，传入父实例作为参数。
     * 每个iframe持有一个独有的gsp实例，其中eventbus共享，ide不共享
     * @return {?}
     */
    createChild() {
        /** @type {?} */
        const child = new GSP(this);
        // child.pEventBus = this.pEventBus;
        // child.pViews = this.pViews;
        // child.pWorkspace = this.pWorkspace;
        return child;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pEventBus;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pUtil;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pWorkspace;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pConfig;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pViews;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pCommands;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pIde;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pCache;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pContext;
}
