// @farris/ide-devkit/esm2015/lib/ideframework/gsp/panel-container.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CompositeDisposable, Emitter } from './event-kit';
export class PanelContainer {
    /**
     * @param {?} viewRegistry
     * @param {?} location
     * @param {?=} dock
     */
    constructor(viewRegistry, location, dock = null) {
        this.viewRegistry = viewRegistry;
        this.location = location;
        this.dock = dock;
        this.panels = [];
        this.emitter = new Emitter();
        this.subscriptions = new CompositeDisposable();
    }
    /**
     * @return {?}
     */
    destroy() {
        for (const panel of this.getPanels()) {
            panel.destroy();
        }
    }
    /**
     * @return {?}
     */
    getElement() {
    }
    /*
      Section: Event Subscription
      */
    /**
     * @param {?} callback
     * @return {?}
     */
    onDidAddPanel(callback) {
        return this.emitter.on('did-add-panel', callback);
    }
    /**
     * @param {?} callback
     * @return {?}
     */
    onDidRemovePanel(callback) {
        return this.emitter.on('did-remove-panel', callback);
    }
    /**
     * @param {?} callback
     * @return {?}
     */
    onDidDestroy(callback) {
        return this.emitter.once('did-destroy', callback);
    }
    /**
     * @return {?}
     */
    getLocation() {
        return this.location;
    }
    /**
     * @return {?}
     */
    isModal() { return this.location === 'modal'; }
    /**
     * @return {?}
     */
    getPanels() { return this.panels.slice(); }
    /**
     * @param {?} panel
     * @return {?}
     */
    addPanel(panel) {
        this.subscriptions.add(panel.onDidDestroy(this.panelDestroyed.bind(this)));
        /** @type {?} */
        const index = this.getPanelIndex(panel);
        if (index === this.panels.length) {
            this.panels.push(panel);
        }
        else {
            this.panels.splice(index, 0, panel);
        }
        this.emitter.emit('did-add-panel', { panel, index });
        return panel;
    }
    /**
     * @param {?} item
     * @return {?}
     */
    panelForItem(item) {
        for (const panel of this.panels) {
            if (panel.getItem() === item) {
                return panel;
            }
        }
        return null;
    }
    /**
     * @param {?} panel
     * @return {?}
     */
    panelDestroyed(panel) {
        /** @type {?} */
        const index = this.panels.indexOf(panel);
        if (index > -1) {
            this.panels.splice(index, 1);
            this.emitter.emit('did-remove-panel', { panel, index });
        }
    }
    /**
     * @param {?} panel
     * @return {?}
     */
    getPanelIndex(panel) {
        /** @type {?} */
        const priority = panel.getPriority();
        if (['bottom', 'right'].find(item => item === this.location)) {
            for (let i = this.panels.length - 1; i >= 0; i--) {
                /** @type {?} */
                const p = this.panels[i];
                if (priority < p.getPriority()) {
                    return i + 1;
                }
            }
            return 0;
        }
        else {
            for (let i = 0; i < this.panels.length; i++) {
                /** @type {?} */
                const p = this.panels[i];
                if (priority < p.getPriority()) {
                    return i;
                }
            }
            return this.panels.length;
        }
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    PanelContainer.prototype.emitter;
    /**
     * @type {?}
     * @private
     */
    PanelContainer.prototype.subscriptions;
    /** @type {?} */
    PanelContainer.prototype.panels;
    /**
     * @type {?}
     * @private
     */
    PanelContainer.prototype.viewRegistry;
    /**
     * @type {?}
     * @private
     */
    PanelContainer.prototype.location;
    /**
     * @type {?}
     * @private
     */
    PanelContainer.prototype.dock;
}
