// @farris/ide-devkit/esm2015/lib/ideframework/gsp/panel.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Emitter } from './event-kit';
/**
 * @record
 */
export function PanelOptions() { }
if (false) {
    /** @type {?} */
    PanelOptions.prototype.item;
    /** @type {?|undefined} */
    PanelOptions.prototype.autoFocus;
    /** @type {?|undefined} */
    PanelOptions.prototype.visible;
    /** @type {?|undefined} */
    PanelOptions.prototype.priority;
    /** @type {?|undefined} */
    PanelOptions.prototype.className;
}
export class Panel {
    /**
     * @param {?} options
     * @param {?} viewRegistry
     */
    constructor(options, viewRegistry) {
        this.destroyed = false;
        this.item = options.item;
        this.autoFocus = options.autoFocus == null ? false : options.autoFocus;
        this.visible = options.visible == null ? true : options.visible;
        this.priority = options.priority == null ? 100 : options.priority;
        this.className = options.className;
        this.viewRegistry = viewRegistry;
        this.emitter = new Emitter();
    }
    // Public: Destroy and remove this panel from the UI.
    /**
     * @return {?}
     */
    destroy() {
        if (this.destroyed) {
            return;
        }
        this.destroyed = true;
        this.hide();
        this.emitter.emit('did-destroy', this);
        return this.emitter.dispose();
    }
    /**
     * @return {?}
     */
    getElement() {
        return this.item.getElement(); // 直接返回item的视图，不再包装一层Panel
    }
    /*
      Section: Event Subscription
      */
    // Public: Invoke the given callback when the pane hidden or shown.
    //
    // * `callback` {Function} to be called when the pane is destroyed.
    //   * `visible` {Boolean} true when the panel has been shown
    //
    // Returns a {Disposable} on which `.dispose()` can be called to unsubscribe.
    /**
     * @param {?} callback
     * @return {?}
     */
    onDidChangeVisible(callback) {
        return this.emitter.on('did-change-visible', callback);
    }
    // Public: Invoke the given callback when the pane is destroyed.
    //
    // * `callback` {Function} to be called when the pane is destroyed.
    //   * `panel` {Panel} this panel
    //
    // Returns a {Disposable} on which `.dispose()` can be called to unsubscribe.
    /**
     * @param {?} callback
     * @return {?}
     */
    onDidDestroy(callback) {
        return this.emitter.once('did-destroy', callback);
    }
    /*
      Section: Panel Details
      */
    // Public: Returns the panel's item.
    /**
     * @return {?}
     */
    getItem() {
        return this.item;
    }
    // Public: Returns a {Number} indicating this panel's priority.
    /**
     * @return {?}
     */
    getPriority() {
        return this.priority;
    }
    /**
     * @return {?}
     */
    getClassName() {
        return this.className;
    }
    // Public: Returns a {Boolean} true when the panel is visible.
    /**
     * @return {?}
     */
    isVisible() {
        return this.visible;
    }
    // Public: Hide this panel
    /**
     * @return {?}
     */
    hide() {
        /** @type {?} */
        const wasVisible = this.visible;
        this.visible = false;
        if (wasVisible) {
            this.emitter.emit('did-change-visible', this.visible);
        }
    }
    // Public: Show this panel
    /**
     * @return {?}
     */
    show() {
        /** @type {?} */
        const wasVisible = this.visible;
        this.visible = true;
        if (!wasVisible) {
            this.emitter.emit('did-change-visible', this.visible);
        }
    }
}
if (false) {
    /** @type {?} */
    Panel.prototype.destroyed;
    /** @type {?} */
    Panel.prototype.item;
    /** @type {?} */
    Panel.prototype.autoFocus;
    /** @type {?} */
    Panel.prototype.visible;
    /** @type {?} */
    Panel.prototype.priority;
    /** @type {?} */
    Panel.prototype.className;
    /** @type {?} */
    Panel.prototype.viewRegistry;
    /**
     * @type {?}
     * @private
     */
    Panel.prototype.emitter;
}
