// @gsp-wf/wf-task-handler/esm2015/lib/entity/farris-user-help-param.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * farris用户帮助参数
 */
export class UserHelpParam {
    constructor() {
        this.readonly = false;
        this.includeJobLeavers = false;
        this.includeStopUser = false;
        this.personnelOrderField = "orderby_ordernum";
        this.userTypeMode = "WfMode";
    }
}
if (false) {
    /** @type {?} */
    UserHelpParam.prototype.readonly;
    /** @type {?} */
    UserHelpParam.prototype.multiSelect;
    /** @type {?} */
    UserHelpParam.prototype.includeJobLeavers;
    /** @type {?} */
    UserHelpParam.prototype.includeCurrentUser;
    /** @type {?} */
    UserHelpParam.prototype.includeStopUser;
    /** @type {?} */
    UserHelpParam.prototype.unSelectedUserIds;
    /** @type {?} */
    UserHelpParam.prototype.orgId;
    /** @type {?} */
    UserHelpParam.prototype.enableChildHierarchy;
    /** @type {?} */
    UserHelpParam.prototype.showTabIds;
    /** @type {?} */
    UserHelpParam.prototype.activeTabId;
    /** @type {?} */
    UserHelpParam.prototype.userIds;
    /** @type {?} */
    UserHelpParam.prototype.personnelOrderField;
    /** @type {?} */
    UserHelpParam.prototype.userTypeMode;
    /** @type {?} */
    UserHelpParam.prototype.filterId;
}
