// @farris/ui-filter-panel/esm5/lib/components/datecomponent/date.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
var PanelDateComponent = /** @class */ (function () {
    function PanelDateComponent() {
        this.searchHandle = new EventEmitter();
        this.submitChange = new EventEmitter();
    }
    Object.defineProperty(PanelDateComponent.prototype, "item", {
        get: /**
         * @return {?}
         */
        function () {
            return this._item;
        },
        set: /**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            var _this = this;
            this._item = data;
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.initDate();
            }), 0);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PanelDateComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    PanelDateComponent.prototype.initDate = /**
     * @return {?}
     */
    function () {
        if (this._item.placeHolder) {
            this.placeholder = this._item.placeHolder || '';
        }
        if (!this._item.hasOwnProperty('value')) {
            this._item.value = { value: '' };
        }
        if (this.item && this._item.value
            && (this._item.value).value) {
            this.date = this._item.value.value;
        }
        else {
            this.date = undefined;
        }
        /** @type {?} */
        var format = this.showType == '2' ? 'yyyy-MM' : (this.showType == '3' ? 'yyyy' : 'yyyy-MM-dd');
        this.dateFormat = this._item.control.format ? this._item.control.format : format;
        this.setSubmitChange();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PanelDateComponent.prototype.onDateChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.returnFormatted) {
            (this._item.value).value = event.returnFormatted;
        }
        else {
            (this._item.value).value = '';
        }
        // this._item.valueText = this._item.value.value;
        this._item.valueText = event.formatted;
        this.setSubmitChange();
        this.searchHandle.emit(event);
    };
    /**
     * @return {?}
     */
    PanelDateComponent.prototype.setSubmitChange = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var val = this._item.value;
        if (Object.keys(val).length < 0 || !val.value) {
            this.submitChange.emit(false);
        }
        else {
            this.submitChange.emit(true);
        }
    };
    PanelDateComponent.decorators = [
        { type: Component, args: [{
                    selector: 'panel-filter-date',
                    template: "\r\n<farris-datepicker\r\n    [(ngModel)]=\"date\"\r\n    [returnFormat]=\"showType == '2'? 'yyyy-MM': (showType == '3'? 'yyyy':'yyyy-MM-dd')\"\r\n    [editable]=\"false\" [dateFormat]=\"dateFormat\"\r\n    (valueChange)=\"onDateChange($event)\"\r\n    [placeholder]=\"placeholder\" [showType]=\"showType\">\r\n</farris-datepicker>"
                }] }
    ];
    /** @nocollapse */
    PanelDateComponent.ctorParameters = function () { return []; };
    PanelDateComponent.propDecorators = {
        item: [{ type: Input }],
        showType: [{ type: Input }],
        searchHandle: [{ type: Output }],
        submitChange: [{ type: Output }]
    };
    return PanelDateComponent;
}());
export { PanelDateComponent };
if (false) {
    /** @type {?} */
    PanelDateComponent.prototype._item;
    /** @type {?} */
    PanelDateComponent.prototype.showType;
    /** @type {?} */
    PanelDateComponent.prototype.searchHandle;
    /** @type {?} */
    PanelDateComponent.prototype.submitChange;
    /** @type {?} */
    PanelDateComponent.prototype.date;
    /** @type {?} */
    PanelDateComponent.prototype.placeholder;
    /** @type {?} */
    PanelDateComponent.prototype.dateFormat;
}
