// @farris/ui-modal/esm5/lib/modal-options.class.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { of } from 'rxjs';
var ModalOptions = /** @class */ (function () {
    function ModalOptions() {
    }
    ModalOptions.decorators = [
        { type: Injectable }
    ];
    return ModalOptions;
}());
export { ModalOptions };
if (false) {
    /**
     *  Includes a modal-backdrop element. Alternatively,
     *  specify static for a backdrop which doesn't close the modal on click.
     * @type {?}
     */
    ModalOptions.prototype.backdrop;
    /**
     * Closes the modal when escape key is pressed.
     * @type {?}
     */
    ModalOptions.prototype.keyboard;
    /** @type {?} */
    ModalOptions.prototype.focus;
    /**
     * Shows the modal when initialized.
     * @type {?}
     */
    ModalOptions.prototype.show;
    /**
     * Ignore the backdrop click
     * @type {?}
     */
    ModalOptions.prototype.ignoreBackdropClick;
    /**
     * Css class for opened modal
     * @type {?}
     */
    ModalOptions.prototype.class;
    /**
     * Toggle animation
     * @type {?}
     */
    ModalOptions.prototype.animated;
    /**
     * Modal data
     * @type {?}
     */
    ModalOptions.prototype.initialState;
    /**
     * 窗口标题
     * @type {?}
     */
    ModalOptions.prototype.title;
    /**
     * 标题前图标
     * @type {?}
     */
    ModalOptions.prototype.iconCls;
    /**
     * 显示头部
     * @type {?}
     */
    ModalOptions.prototype.showHeader;
    /**
     * 显示关闭按钮
     * @type {?}
     */
    ModalOptions.prototype.showCloseButton;
    /**
     * 显示最大化按钮
     * @type {?}
     */
    ModalOptions.prototype.showMaxButton;
    /**
     * 显示最小化按钮
     * @type {?}
     */
    ModalOptions.prototype.showMinButton;
    /**
     * 窗口宽度
     * @type {?}
     */
    ModalOptions.prototype.width;
    /**
     * 窗口高度
     * @type {?}
     */
    ModalOptions.prototype.height;
    /** @type {?} */
    ModalOptions.prototype.minWidth;
    /** @type {?} */
    ModalOptions.prototype.minHeight;
    /** @type {?} */
    ModalOptions.prototype.maxWidth;
    /** @type {?} */
    ModalOptions.prototype.maxHeight;
    /**
     * 窗口标题栏高度
     * @type {?}
     */
    ModalOptions.prototype.dialogHeaderHeight;
    /**
     * 窗口按钮区高度
     * @type {?}
     */
    ModalOptions.prototype.dialogFooterHeight;
    /** @type {?} */
    ModalOptions.prototype.dialogFooterStyles;
    /**
     * 显示按钮区
     * @type {?}
     */
    ModalOptions.prototype.showButtons;
    /**
     * 按钮对齐方式
     * @type {?}
     */
    ModalOptions.prototype.buttonAlign;
    /**
     * 按钮列表模板
     * @type {?}
     */
    ModalOptions.prototype.buttons;
    /**
     * 启用拖拽功能
     * @type {?}
     */
    ModalOptions.prototype.draggable;
    /**
     * 启用调整尺寸大小
     * @type {?}
     */
    ModalOptions.prototype.resizable;
    /**
     * 是否启用滚动条
     * @type {?}
     */
    ModalOptions.prototype.enableScroll;
    /** @type {?} */
    ModalOptions.prototype.beforeOpen;
    /** @type {?} */
    ModalOptions.prototype.beforeClose;
    /** @type {?} */
    ModalOptions.prototype.opened;
    /** @type {?} */
    ModalOptions.prototype.closed;
    /** @type {?} */
    ModalOptions.prototype.okText;
    /** @type {?} */
    ModalOptions.prototype.cancelText;
    /**
     * 弹窗类型：default，弹出组件、模板等；iframe,弹出iframe; ngFactory, 弹出ngFactory;
     * @type {?}
     */
    ModalOptions.prototype.dialogType;
    /** @type {?} */
    ModalOptions.prototype.fitContent;
    /** @type {?} */
    ModalOptions.prototype.areaResponse;
    /** @type {?} */
    ModalOptions.prototype.exception;
    /** @type {?} */
    ModalOptions.prototype.modeless;
}
/**
 * @record
 */
export function DialogButton() { }
if (false) {
    /** @type {?} */
    DialogButton.prototype.text;
    /** @type {?|undefined} */
    DialogButton.prototype.cls;
    /** @type {?|undefined} */
    DialogButton.prototype.iconCls;
    /** @type {?|undefined} */
    DialogButton.prototype.disable;
    /** @type {?|undefined} */
    DialogButton.prototype.defaultFocus;
    /** @type {?|undefined} */
    DialogButton.prototype.handle;
    /** @type {?|undefined} */
    DialogButton.prototype.display;
    /** @type {?|undefined} */
    DialogButton.prototype.tipsEnable;
    /** @type {?|undefined} */
    DialogButton.prototype.tipsText;
}
/** @type {?} */
export var modalConfigDefaults = {
    backdrop: 'static',
    keyboard: true,
    focus: true,
    show: false,
    ignoreBackdropClick: true,
    class: '',
    animated: true,
    initialState: {},
    width: 400,
    height: 350,
    minHeight: null,
    minWidth: null,
    maxHeight: null,
    maxWidth: null,
    title: 'Title is here.',
    showCloseButton: true,
    showMaxButton: true,
    showButtons: true,
    buttonAlign: 'center',
    draggable: true,
    resizable: true,
    enableScroll: true,
    beforeOpen: (/**
     * @return {?}
     */
    function () { return of(true); }),
    beforeClose: (/**
     * @return {?}
     */
    function () { return of(true); }),
    okText: '确定',
    cancelText: '取消',
    dialogHeaderHeight: 38,
    dialogFooterHeight: 50,
    dialogType: 'default',
    showHeader: true,
    fitContent: false,
    areaResponse: false
};
/** @type {?} */
export var CLASS_NAME = {
    SCROLLBAR_MEASURER: 'modal-scrollbar-measure',
    BACKDROP: 'modal-backdrop',
    OPEN: 'modal-open',
    FADE: 'fade',
    IN: 'in',
    // bs3
    SHOW: 'show' // bs4
};
/** @type {?} */
export var SELECTOR = {
    DIALOG: '.modal-dialog',
    DATA_TOGGLE: '[data-toggle="modal"]',
    DATA_DISMISS: '[data-dismiss="modal"]',
    FIXED_CONTENT: '.navbar-fixed-top, .navbar-fixed-bottom, .is-fixed'
};
/** @type {?} */
export var TRANSITION_DURATIONS = {
    MODAL: 300,
    BACKDROP: 150
};
/** @type {?} */
export var DISMISS_REASONS = {
    BACKRDOP: 'backdrop-click',
    ESC: 'esc'
};
