// @farris/ui-modal/esm5/lib/modal-button.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Injector, ElementRef, Input } from '@angular/core';
var ModalButtonDirective = /** @class */ (function () {
    function ModalButtonDirective(injector, el) {
        this.injector = injector;
        this.el = el;
        this.visable = true;
    }
    /**
     * @param {?} disable
     * @return {?}
     */
    ModalButtonDirective.prototype.setDisabled = /**
     * @param {?} disable
     * @return {?}
     */
    function (disable) {
        this.el.nativeElement.disabled = disable;
    };
    /**
     * @return {?}
     */
    ModalButtonDirective.prototype.show = /**
     * @return {?}
     */
    function () {
        this.visable = true;
    };
    /**
     * @return {?}
     */
    ModalButtonDirective.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.visable = false;
    };
    ModalButtonDirective.decorators = [
        { type: Directive, args: [{ selector: '[modal-button]' },] }
    ];
    /** @nocollapse */
    ModalButtonDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: ElementRef }
    ]; };
    ModalButtonDirective.propDecorators = {
        visable: [{ type: Input }]
    };
    return ModalButtonDirective;
}());
export { ModalButtonDirective };
if (false) {
    /** @type {?} */
    ModalButtonDirective.prototype.visable;
    /**
     * @type {?}
     * @private
     */
    ModalButtonDirective.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    ModalButtonDirective.prototype.el;
}
