// @edp-aif/common-api/esm2015/lib/ExtensionMethod.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from './BpmnModelJsonHelper';
export class ExtensionMethod {
    /**
     * @param {?} obj
     * @param {?} attributeName
     * @return {?}
     */
    static getAttributeValue(obj, attributeName) {
        return BpmnModelJsonHelper.GetAttributeValue(obj, attributeName);
    }
    /**
     * @param {?} obj
     * @param {?} attributeName
     * @return {?}
     */
    static getAttributeValue4Boolean(obj, attributeName) {
        return BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, attributeName);
    }
    /**
     * @param {?} obj
     * @param {?} attributeName
     * @return {?}
     */
    static getAttributeValue4Decimal(obj, attributeName) {
        return BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, attributeName);
    }
    /**
     * @param {?} obj
     * @param {?} attributeName
     * @return {?}
     */
    static getAttributeValue4String(obj, attributeName) {
        return BpmnModelJsonHelper.GetAttributeValueAsString(obj, attributeName);
    }
    /**
     * @param {?} obj
     * @param {?} attributeName
     * @param {?} value
     * @return {?}
     */
    static setAttributeValue(obj, attributeName, value) {
        BpmnModelJsonHelper.SetAttributeValue(obj, attributeName, value);
    }
}
