// @farris/appointment-calendar/esm5/lib/dblclick.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Directive, ElementRef, HostListener, Input, Output } from '@angular/core';
var ItemDblClickDirective = /** @class */ (function () {
    function ItemDblClickDirective(el) {
        this.el = el;
        this.dblclickHandler = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ItemDblClickDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} event
     * @return {?}
     */
    ItemDblClickDirective.prototype.ondblclick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.stopPropagation();
        this.dblclickHandler.emit({ item: this.data, $event: event });
    };
    ItemDblClickDirective.decorators = [
        { type: Directive, args: [{ selector: '[item-dblclick]' },] }
    ];
    /** @nocollapse */
    ItemDblClickDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    ItemDblClickDirective.propDecorators = {
        data: [{ type: Input, args: ['item-dblclick',] }],
        dblclickHandler: [{ type: Output }],
        ondblclick: [{ type: HostListener, args: ['dblclick', ['$event'],] }]
    };
    return ItemDblClickDirective;
}());
export { ItemDblClickDirective };
if (false) {
    /** @type {?} */
    ItemDblClickDirective.prototype.data;
    /** @type {?} */
    ItemDblClickDirective.prototype.dblclickHandler;
    /** @type {?} */
    ItemDblClickDirective.prototype.el;
}
