// @gsp-cmp/common-component/esm5/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of component
 */
/**
 * define
 */
export { GspComponent } from './define/gsp-component';
export { StructureReferType } from './define/data-type/structure-refer-type';
export { RefStructureInfo } from './define/ref-structure-info';
/**
 * Constant
 */
export { GspComponentConstant } from './define/gsp-component-constant';
/**
 * convertor
 */
export { GspComponentConvertor } from "./convertor/gsp-component-convertor";
export { RefStructureUtil } from "./convertor/util/ref-structure-util";
export { StructureReferTypeConvertor } from "./convertor/data-type/structure-refer-type-convertor";
/**
 * default serialize context
 */
export { DefaultGspComponentSerializeContext } from './serialize-context/default/default-gsp-component-serialize-context';
export { DefaultGspComponentDataTypeSerializerFactory } from './serialize-context/default/factory/default-gsp-component-data-type-serializer-factory';
export { DefaultGspComponentSerializerFactory } from './serialize-context/default/factory/default-gsp-component-serializer-factory';
export { DefaultGspComponentDataTypeSerializer } from './serialize-context/default/serializer/default-gsp-component-data-type-serializer';
export { DefaultGspComponentSerializer } from './serialize-context/default/serializer/default-gsp-component-serializer';
export { CommonComponentModule } from './common-component.module';
export { CommonComponentService } from './service/common-component.service';
