// @progress/kendo-angular-dialog/dist/es/dialog/dialog-titlebar.component.js
import { Component, EventEmitter, HostBinding, Output, Input, ElementRef } from '@angular/core';
/**
 * @hidden
 */
var DialogTitleBarComponent = /** @class */ (function () {
    function DialogTitleBarComponent(hostElement) {
        this.hostElement = hostElement;
        this.close = new EventEmitter();
    }
    Object.defineProperty(DialogTitleBarComponent.prototype, "className", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    DialogTitleBarComponent.prototype.ngAfterViewChecked = function () {
        var element = this.hostElement.nativeElement;
        element.setAttribute('id', this.id);
    };
    DialogTitleBarComponent.prototype.onCloseClick = function (e) {
        e.preventDefault();
        this.close.emit();
    };
    DialogTitleBarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'kendo-dialog-titlebar',
                    template: "\n    <div class=\"k-window-title k-dialog-title\">\n      <ng-content></ng-content>\n    </div>\n    <div class=\"k-window-actions k-dialog-actions\">\n      <a href=\"#\" role=\"button\"\n         aria-label=\"Close\"\n         class=\"k-button k-bare k-button-icon k-window-action k-dialog-action k-dialog-close\"\n         (click)=\"onCloseClick($event)\">\n        <span class=\"k-icon k-i-x\"></span>\n      </a>\n    </div>\n  "
                },] },
    ];
    /** @nocollapse */
    DialogTitleBarComponent.ctorParameters = function () { return [
        { type: ElementRef, },
    ]; };
    DialogTitleBarComponent.propDecorators = {
        'close': [{ type: Output },],
        'id': [{ type: Input },],
        'className': [{ type: HostBinding, args: ['class.k-window-titlebar',] }, { type: HostBinding, args: ['class.k-dialog-titlebar',] }, { type: HostBinding, args: ['class.k-header',] },],
    };
    return DialogTitleBarComponent;
}());
export { DialogTitleBarComponent };
