// @edp-aif/common-api/esm5/lib/billkit/BillData.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '../BpmnModelJsonHelper';
import { KitBill } from './KitBill';
import { BillDataTable } from './BillDataTable';
var BillData = /** @class */ (function (_super) {
    tslib_1.__extends(BillData, _super);
    function BillData() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    BillData.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.mainTable = null;
    };
    /**
     * @return {?}
     */
    BillData.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mainTable", this.mainTable);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BillData.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.mainTable = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "mainTable", BillDataTable);
        this.ApplyRelationship();
    };
    /*
     * 将数据集、主表、子表、行、字段之间的隶属关系应用到各个对象上。
     * */
    /*
         * 将数据集、主表、子表、行、字段之间的隶属关系应用到各个对象上。
         * */
    /**
     * @return {?}
     */
    BillData.prototype.ApplyRelationship = /*
         * 将数据集、主表、子表、行、字段之间的隶属关系应用到各个对象上。
         * */
    /**
     * @return {?}
     */
    function () {
        this.Apply2Table(this, null, this.mainTable);
        return this;
    };
    /**
     * @private
     * @param {?} attachedBillData
     * @param {?} attachedRow
     * @param {?} table
     * @return {?}
     */
    BillData.prototype.Apply2Table = /**
     * @private
     * @param {?} attachedBillData
     * @param {?} attachedRow
     * @param {?} table
     * @return {?}
     */
    function (attachedBillData, attachedRow, table) {
        var e_1, _a;
        table.attachedBillData = attachedBillData;
        table.attachedRow = attachedRow;
        try {
            for (var _b = tslib_1.__values(table.GetRowList()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var row = _c.value;
                this.Apply2Row(table, row);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    /**
     * @private
     * @param {?} attachedTable
     * @param {?} row
     * @return {?}
     */
    BillData.prototype.Apply2Row = /**
     * @private
     * @param {?} attachedTable
     * @param {?} row
     * @return {?}
     */
    function (attachedTable, row) {
        var e_2, _a, e_3, _b;
        row.attachedTable = attachedTable;
        try {
            for (var _c = tslib_1.__values(row.GetFieldList()), _d = _c.next(); !_d.done; _d = _c.next()) {
                var field = _d.value;
                this.Apply2Field(row, field);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_2) throw e_2.error; }
        }
        try {
            for (var _e = tslib_1.__values(row.GetChildTableList()), _f = _e.next(); !_f.done; _f = _e.next()) {
                var childTable = _f.value;
                this.Apply2Table(attachedTable.attachedBillData, row, childTable);
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
            }
            finally { if (e_3) throw e_3.error; }
        }
    };
    /**
     * @private
     * @param {?} attachedRow
     * @param {?} field
     * @return {?}
     */
    BillData.prototype.Apply2Field = /**
     * @private
     * @param {?} attachedRow
     * @param {?} field
     * @return {?}
     */
    function (attachedRow, field) {
        var e_4, _a;
        field.attachedRow = attachedRow;
        try {
            for (var _b = tslib_1.__values(field.GetSubFieldList()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var subField = _c.value;
                this.Apply2SubField(field, subField);
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_4) throw e_4.error; }
        }
    };
    /**
     * @private
     * @param {?} attachedField
     * @param {?} field
     * @return {?}
     */
    BillData.prototype.Apply2SubField = /**
     * @private
     * @param {?} attachedField
     * @param {?} field
     * @return {?}
     */
    function (attachedField, field) {
        var e_5, _a;
        field.attachedField = field;
        try {
            for (var _b = tslib_1.__values(field.GetSubFieldList()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var subField = _c.value;
                this.Apply2SubField(field, subField);
            }
        }
        catch (e_5_1) { e_5 = { error: e_5_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_5) throw e_5.error; }
        }
    };
    return BillData;
}(KitBill));
export { BillData };
if (false) {
    /** @type {?} */
    BillData.prototype.mainTable;
}
