// @edp-aif/common-api/esm5/lib/billkit/BillSchema.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '../BpmnModelJsonHelper';
import { KitBill } from './KitBill';
import { BillSchemaTable } from './BillSchemaTable';
var BillSchema = /** @class */ (function (_super) {
    tslib_1.__extends(BillSchema, _super);
    function BillSchema() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    BillSchema.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.mainTable = null;
    };
    /**
     * @return {?}
     */
    BillSchema.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mainTable", this.mainTable);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BillSchema.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.mainTable = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "mainTable", BillSchemaTable);
        this.ApplyRelationship();
    };
    /*
     * 将架构、主表、子表、列、字段之间的隶属关系应用到各个对象上。
     * */
    /*
         * 将架构、主表、子表、列、字段之间的隶属关系应用到各个对象上。
         * */
    /**
     * @return {?}
     */
    BillSchema.prototype.ApplyRelationship = /*
         * 将架构、主表、子表、列、字段之间的隶属关系应用到各个对象上。
         * */
    /**
     * @return {?}
     */
    function () {
        this.Apply2Table(this, null, this.mainTable);
        return this;
    };
    /**
     * @private
     * @param {?} attachedBillSchema
     * @param {?} attachedTable
     * @param {?} table
     * @return {?}
     */
    BillSchema.prototype.Apply2Table = /**
     * @private
     * @param {?} attachedBillSchema
     * @param {?} attachedTable
     * @param {?} table
     * @return {?}
     */
    function (attachedBillSchema, attachedTable, table) {
        var e_1, _a, e_2, _b;
        table.attachedBillSchema = attachedBillSchema;
        table.attachedTable = attachedTable;
        try {
            for (var _c = tslib_1.__values(table.getFieldList()), _d = _c.next(); !_d.done; _d = _c.next()) {
                var field = _d.value;
                this.Apply2Field(table, field);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        try {
            for (var _e = tslib_1.__values(table.GetChildTableList()), _f = _e.next(); !_f.done; _f = _e.next()) {
                var childTable = _f.value;
                this.Apply2Table(table.attachedBillSchema, table, childTable);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    /**
     * @private
     * @param {?} attachedTable
     * @param {?} field
     * @return {?}
     */
    BillSchema.prototype.Apply2Field = /**
     * @private
     * @param {?} attachedTable
     * @param {?} field
     * @return {?}
     */
    function (attachedTable, field) {
        var e_3, _a;
        field.attachedTable = attachedTable;
        try {
            for (var _b = tslib_1.__values(field.GetSubFieldList()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var subField = _c.value;
                this.Apply2SubField(field, subField);
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
    };
    /**
     * @private
     * @param {?} attachedField
     * @param {?} field
     * @return {?}
     */
    BillSchema.prototype.Apply2SubField = /**
     * @private
     * @param {?} attachedField
     * @param {?} field
     * @return {?}
     */
    function (attachedField, field) {
        var e_4, _a;
        field.attachedField = attachedField;
        try {
            for (var _b = tslib_1.__values(field.GetSubFieldList()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var subField = _c.value;
                this.Apply2SubField(field, subField);
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_4) throw e_4.error; }
        }
    };
    return BillSchema;
}(KitBill));
export { BillSchema };
if (false) {
    /** @type {?} */
    BillSchema.prototype.mainTable;
}
