// @edp-aif/common-api/esm5/lib/billkit/BillDataTable.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '../BpmnModelJsonHelper';
import { BillDataRow } from './BillDataRow';
import { BillDataRows } from './BillDataRows';
import { KitTable } from './KitTable';
var BillDataTable = /** @class */ (function (_super) {
    tslib_1.__extends(BillDataTable, _super);
    function BillDataTable() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.rows = new BillDataRows();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    BillDataTable.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.attachedBillData = null;
        this.attachedRow = null;
        this.rows.Clear();
    };
    /**
     * @return {?}
     */
    BillDataTable.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        /** @type {?} */
        var dataRows = [];
        try {
            for (var _b = tslib_1.__values(this.rows.GetList()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var value = _c.value;
                dataRows.push(value.ConvertToJsonObject());
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        BpmnModelJsonHelper.SetAttributeValue(obj, "rows", dataRows);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BillDataTable.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var e_2, _a;
        _super.prototype.LoadFromJsonObject.call(this, obj);
        /** @type {?} */
        var dataRows = BpmnModelJsonHelper.GetAttributeValue(obj, "rows");
        try {
            for (var dataRows_1 = tslib_1.__values(dataRows), dataRows_1_1 = dataRows_1.next(); !dataRows_1_1.done; dataRows_1_1 = dataRows_1.next()) {
                var item = dataRows_1_1.value;
                /** @type {?} */
                var value = new BillDataRow();
                value.LoadFromJsonObject(item);
                this.rows.Add(value);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (dataRows_1_1 && !dataRows_1_1.done && (_a = dataRows_1.return)) _a.call(dataRows_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    /**
     * @param {?} row
     * @return {?}
     */
    BillDataTable.prototype.AddRow = /**
     * @param {?} row
     * @return {?}
     */
    function (row) {
        this.rows.Add(row);
    };
    /**
     * @return {?}
     */
    BillDataTable.prototype.GetRowMap = /**
     * @return {?}
     */
    function () {
        return this.rows.GetMap();
    };
    /**
     * @return {?}
     */
    BillDataTable.prototype.GetRowList = /**
     * @return {?}
     */
    function () {
        return this.rows.GetList();
    };
    return BillDataTable;
}(KitTable));
export { BillDataTable };
if (false) {
    /** @type {?} */
    BillDataTable.prototype.rows;
    /** @type {?} */
    BillDataTable.prototype.attachedBillData;
    /** @type {?} */
    BillDataTable.prototype.attachedRow;
}
