// @edp-aif/common-api/esm5/lib/billkit/BillSchemaFields.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BillSchemaFields = /** @class */ (function () {
    function BillSchemaFields() {
        this._map = {};
        this._list = new Array();
    }
    Object.defineProperty(BillSchemaFields.prototype, "map", {
        get: /**
         * @return {?}
         */
        function () {
            return this._map;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BillSchemaFields.prototype, "list", {
        get: /**
         * @return {?}
         */
        function () {
            return this._list;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    BillSchemaFields.prototype.GetMap = /**
     * @return {?}
     */
    function () {
        return this._map;
    };
    /**
     * @return {?}
     */
    BillSchemaFields.prototype.GetList = /**
     * @return {?}
     */
    function () {
        return this._list;
    };
    /**
     * @param {?} field
     * @return {?}
     */
    BillSchemaFields.prototype.Add = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        this._map[field.fieldName] = field;
        this._list.push(field);
    };
    /**
     * @param {?} field
     * @return {?}
     */
    BillSchemaFields.prototype.Remove = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        /** @type {?} */
        var bak = this._map[field.fieldName];
        delete this._map[field.fieldName];
        for (var i = 0; i < this._list.length; i++) {
            if (this._list[i].fieldName == bak.fieldName) {
                this._list.splice(i, 1);
                return true;
            }
        }
        return false;
    };
    /**
     * @return {?}
     */
    BillSchemaFields.prototype.Clear = /**
     * @return {?}
     */
    function () {
        this._map = {};
        this._list.length = 0;
    };
    return BillSchemaFields;
}());
export { BillSchemaFields };
if (false) {
    /**
     * @type {?}
     * @private
     */
    BillSchemaFields.prototype._map;
    /**
     * @type {?}
     * @private
     */
    BillSchemaFields.prototype._list;
}
