// @edp-aif/common-api/esm5/lib/billkit/BillDataTables.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BillDataTables = /** @class */ (function () {
    function BillDataTables() {
        this._map = {};
        this._list = new Array();
    }
    Object.defineProperty(BillDataTables.prototype, "map", {
        get: /**
         * @return {?}
         */
        function () {
            return this._map;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BillDataTables.prototype, "list", {
        get: /**
         * @return {?}
         */
        function () {
            return this._list;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    BillDataTables.prototype.GetMap = /**
     * @return {?}
     */
    function () {
        return this._map;
    };
    /**
     * @return {?}
     */
    BillDataTables.prototype.GetList = /**
     * @return {?}
     */
    function () {
        return this._list;
    };
    /**
     * @param {?} table
     * @return {?}
     */
    BillDataTables.prototype.Add = /**
     * @param {?} table
     * @return {?}
     */
    function (table) {
        this._map[table.tableName] = table;
        this._list.push(table);
    };
    /**
     * @param {?} table
     * @return {?}
     */
    BillDataTables.prototype.Remove = /**
     * @param {?} table
     * @return {?}
     */
    function (table) {
        /** @type {?} */
        var bak = this._map[table.tableName];
        delete this._map[table.tableName];
        for (var i = 0; i < this._list.length; i++) {
            if (this._list[i].tableName == bak.tableName) {
                this._list.splice(i, 1);
                return true;
            }
        }
        return false;
    };
    /**
     * @return {?}
     */
    BillDataTables.prototype.Clear = /**
     * @return {?}
     */
    function () {
        this._map = {};
        this._list.length = 0;
    };
    return BillDataTables;
}());
export { BillDataTables };
if (false) {
    /**
     * @type {?}
     * @private
     */
    BillDataTables.prototype._map;
    /**
     * @type {?}
     * @private
     */
    BillDataTables.prototype._list;
}
