// @edp-aif/common-api/esm5/lib/billkit/BillDataField.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BillDataFieldBase } from './BillDataFieldBase';
import { BillDataSubFields } from './BillDataSubFields';
import { BpmnModelJsonHelper } from '../BpmnModelJsonHelper';
import { BillDataSubField } from './BillDataSubField';
var BillDataField = /** @class */ (function (_super) {
    tslib_1.__extends(BillDataField, _super);
    function BillDataField() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.subFields = new BillDataSubFields();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    BillDataField.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.subFields.Clear();
        this.attachedRow = null;
    };
    /**
     * @return {?}
     */
    BillDataField.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        if (this.subFields.GetList().length > 0) {
            /** @type {?} */
            var subFields = [];
            try {
                for (var _b = tslib_1.__values(this.subFields.GetList()), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var value = _c.value;
                    subFields.push(value.ConvertToJsonObject());
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            BpmnModelJsonHelper.SetAttributeValue(obj, "subFields", subFields);
        }
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BillDataField.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var e_2, _a;
        _super.prototype.LoadFromJsonObject.call(this, obj);
        /** @type {?} */
        var subFields = BpmnModelJsonHelper.GetAttributeValue(obj, "subFields");
        if (subFields != null) {
            try {
                for (var subFields_1 = tslib_1.__values(subFields), subFields_1_1 = subFields_1.next(); !subFields_1_1.done; subFields_1_1 = subFields_1.next()) {
                    var item = subFields_1_1.value;
                    /** @type {?} */
                    var value = new BillDataSubField();
                    value.LoadFromJsonObject(item);
                    this.subFields.Add(value);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (subFields_1_1 && !subFields_1_1.done && (_a = subFields_1.return)) _a.call(subFields_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
    };
    /**
     * @param {?} field
     * @return {?}
     */
    BillDataField.prototype.AddSubField = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        this.subFields.Add(field);
    };
    /**
     * @return {?}
     */
    BillDataField.prototype.GetSubFieldMap = /**
     * @return {?}
     */
    function () {
        return this.subFields.GetMap();
    };
    /**
     * @return {?}
     */
    BillDataField.prototype.GetSubFieldList = /**
     * @return {?}
     */
    function () {
        return this.subFields.GetList();
    };
    return BillDataField;
}(BillDataFieldBase));
export { BillDataField };
if (false) {
    /** @type {?} */
    BillDataField.prototype.subFields;
    /** @type {?} */
    BillDataField.prototype.attachedRow;
}
