// @edp-aif/common-api/esm5/lib/billkit/BillDataRows.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BillDataRows = /** @class */ (function () {
    function BillDataRows() {
        this._map = {};
        this._list = new Array();
    }
    Object.defineProperty(BillDataRows.prototype, "map", {
        get: /**
         * @return {?}
         */
        function () {
            return this._map;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BillDataRows.prototype, "list", {
        get: /**
         * @return {?}
         */
        function () {
            return this._list;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} row
     * @return {?}
     */
    BillDataRows.prototype.Add = /**
     * @param {?} row
     * @return {?}
     */
    function (row) {
        this._map[row.dataId] = row;
        this._list.push(row);
    };
    /**
     * @param {?} row
     * @return {?}
     */
    BillDataRows.prototype.Remove = /**
     * @param {?} row
     * @return {?}
     */
    function (row) {
        /** @type {?} */
        var bak = this._map[row.dataId];
        delete this._map[row.dataId];
        for (var i = 0; i < this._list.length; i++) {
            if (this._list[i].dataId == bak.dataId) {
                this._list.splice(i, 1);
                return true;
            }
        }
        return false;
    };
    /**
     * @return {?}
     */
    BillDataRows.prototype.Clear = /**
     * @return {?}
     */
    function () {
        this._map = {};
        this._list.length = 0;
    };
    /**
     * @return {?}
     */
    BillDataRows.prototype.GetMap = /**
     * @return {?}
     */
    function () {
        return this._map;
    };
    /**
     * @return {?}
     */
    BillDataRows.prototype.GetList = /**
     * @return {?}
     */
    function () {
        return this._list;
    };
    return BillDataRows;
}());
export { BillDataRows };
if (false) {
    /**
     * @type {?}
     * @private
     */
    BillDataRows.prototype._map;
    /**
     * @type {?}
     * @private
     */
    BillDataRows.prototype._list;
}
