// @edp-aif/common-api/esm5/lib/billkit/KitBill.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '../BpmnModelJsonHelper';
import { JsonConvertableImpl } from '../JsonConvertableImpl';
/**
 * @abstract
 */
var /**
 * @abstract
 */
KitBill = /** @class */ (function (_super) {
    tslib_1.__extends(KitBill, _super);
    function KitBill() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    KitBill.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.schemaCategoryId = null;
        this.schemaId = null;
        this.schemaCaption = null;
    };
    /**
     * @return {?}
     */
    KitBill.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaCategoryId", this.schemaCategoryId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaId", this.schemaId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaCaption", this.schemaCaption);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    KitBill.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.schemaCategoryId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaCategoryId");
        this.schemaId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaId");
        this.schemaCaption = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaCaption");
    };
    return KitBill;
}(JsonConvertableImpl));
/**
 * @abstract
 */
export { KitBill };
if (false) {
    /** @type {?} */
    KitBill.prototype.schemaCategoryId;
    /** @type {?} */
    KitBill.prototype.schemaId;
    /** @type {?} */
    KitBill.prototype.schemaCaption;
}
