// @edp-aif/common-api/esm5/lib/billkit/KitFieldInstance.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { KitField } from './KitField';
import { DataTypeStoreImpl } from './DataTypeStoreImpl';
import { ExtensionMethod } from '../ExtensionMethod';
/**
 * @abstract
 */
var /**
 * @abstract
 */
KitFieldInstance = /** @class */ (function (_super) {
    tslib_1.__extends(KitFieldInstance, _super);
    function KitFieldInstance(fieldName, fieldValue) {
        var _this = _super.call(this) || this;
        _this.fieldStore = new DataTypeStoreImpl();
        _this.fieldName = fieldName;
        _this.SetFieldValue(fieldValue);
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    KitFieldInstance.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.fieldStore = new DataTypeStoreImpl();
    };
    /**
     * @return {?}
     */
    KitFieldInstance.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        if ("base.string" != this.fieldStore.converterName) {
            ExtensionMethod.setAttributeValue(obj, "conv", this.fieldStore.converterName);
        }
        if (this.fieldStore.textValue != null) {
            ExtensionMethod.setAttributeValue(obj, "value", this.fieldStore.textValue);
        }
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    KitFieldInstance.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        /** @type {?} */
        var conv = ExtensionMethod.getAttributeValue4String(obj, 'conv');
        if (conv == null) {
            conv = "base.string";
        }
        this.fieldStore.converterName = conv;
        /** @type {?} */
        var text = ExtensionMethod.getAttributeValue4String(obj, "value");
        if (text != null) {
            this.fieldStore.textValue = text;
        }
    };
    /**
     * @return {?}
     */
    KitFieldInstance.prototype.GetFieldValue = /**
     * @return {?}
     */
    function () {
        return this.fieldStore.dataValue;
    };
    /**
     * @param {?} fieldValue
     * @return {?}
     */
    KitFieldInstance.prototype.SetFieldValue = /**
     * @param {?} fieldValue
     * @return {?}
     */
    function (fieldValue) {
        this.fieldStore.dataValue = fieldValue;
    };
    /**
     * @return {?}
     */
    KitFieldInstance.prototype.GetTextValue = /**
     * @return {?}
     */
    function () {
        return this.fieldStore.textValue;
    };
    return KitFieldInstance;
}(KitField));
/**
 * @abstract
 */
export { KitFieldInstance };
if (false) {
    /**
     * @type {?}
     * @private
     */
    KitFieldInstance.prototype.fieldStore;
}
