// @farris/ui-datagrid/esm2015/lib/components/body/datagrid-grouprow.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Injector, Renderer2, ElementRef } from '@angular/core';
import { CUSTOM_GROUP_ROW_CLS } from '../../types/constant';
import { DatagridComponent } from './../../datagrid.component';
export class DatagridGroupRowDirective {
    /**
     * @param {?} injector
     * @param {?} render
     * @param {?} el
     * @param {?} dg
     */
    constructor(injector, render, el, dg) {
        this.injector = injector;
        this.render = render;
        this.el = el;
        this.dg = dg;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        setTimeout((/**
         * @return {?}
         */
        () => {
            if (this.dg && this.dg.groupStyler) {
                /** @type {?} */
                const tr = this.el.nativeElement;
                /** @type {?} */
                const _cls = Array.from(tr.classList).find((/**
                 * @param {?} n
                 * @return {?}
                 */
                (n) => n.indexOf(CUSTOM_GROUP_ROW_CLS) > -1));
                if (_cls) {
                    tr.classList.remove(_cls);
                }
                /** @type {?} */
                const styler = this.dg.groupStyler(this.groupRow);
                if (!styler) {
                    return;
                }
                if (!tr.id) {
                    tr.id = 'group-tr_' + this.groupRow.id;
                }
                /** @type {?} */
                const str = CUSTOM_GROUP_ROW_CLS + this.dg.idService.encrypt(JSON.stringify(styler));
                this.render.addClass(tr, str);
                this.dg.renderCustomStyle2(styler, tr, str);
                // this.dg.renderCustomStyle(styler, tr);
            }
        }));
    }
}
DatagridGroupRowDirective.decorators = [
    { type: Directive, args: [{
                selector: '[group-row]',
            },] }
];
/** @nocollapse */
DatagridGroupRowDirective.ctorParameters = () => [
    { type: Injector },
    { type: Renderer2 },
    { type: ElementRef },
    { type: DatagridComponent }
];
DatagridGroupRowDirective.propDecorators = {
    groupRow: [{ type: Input, args: ['group-row',] }]
};
if (false) {
    /** @type {?} */
    DatagridGroupRowDirective.prototype.groupRow;
    /**
     * @type {?}
     * @private
     */
    DatagridGroupRowDirective.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    DatagridGroupRowDirective.prototype.render;
    /**
     * @type {?}
     * @private
     */
    DatagridGroupRowDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    DatagridGroupRowDirective.prototype.dg;
}
