// @farris/ui-datagrid/esm2015/lib/components/body/datagrid-group-footer-styler.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Renderer2, Injector, Input } from '@angular/core';
import { DatagridComponent } from '../../datagrid.component';
export class GroupFooterRowStylerDirective {
    /**
     * @param {?} el
     * @param {?} render
     * @param {?} inject
     * @param {?} dg
     */
    constructor(el, render, inject, dg) {
        this.el = el;
        this.render = render;
        this.inject = inject;
        this.dg = dg;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.buildCustomGroupFooterRowStyle(this.rowData);
    }
    /**
     * @private
     * @param {?} rowData
     * @return {?}
     */
    buildCustomGroupFooterRowStyle(rowData) {
        if (this.dg.groupFooterStyler) {
            /** @type {?} */
            const cs = this.dg.dgs.getCustomStyle(this.dg.groupFooterStyler, [rowData]);
            if (cs) {
                this.dg.renderCustomCls(cs.cls, this.el.nativeElement);
                if (cs.style) {
                    Object.keys(cs.style).forEach((/**
                     * @param {?} n
                     * @return {?}
                     */
                    n => {
                        this.el.nativeElement.style[n] = cs.style[n];
                    }));
                }
            }
        }
    }
}
GroupFooterRowStylerDirective.decorators = [
    { type: Directive, args: [{
                selector: '[group-footer-row-styler]',
            },] }
];
/** @nocollapse */
GroupFooterRowStylerDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: Injector },
    { type: DatagridComponent }
];
GroupFooterRowStylerDirective.propDecorators = {
    column: [{ type: Input }],
    rowData: [{ type: Input, args: ['group-footer-row-styler',] }],
    rowIndex: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    GroupFooterRowStylerDirective.prototype.column;
    /** @type {?} */
    GroupFooterRowStylerDirective.prototype.rowData;
    /** @type {?} */
    GroupFooterRowStylerDirective.prototype.rowIndex;
    /**
     * @type {?}
     * @private
     */
    GroupFooterRowStylerDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    GroupFooterRowStylerDirective.prototype.render;
    /**
     * @type {?}
     * @private
     */
    GroupFooterRowStylerDirective.prototype.inject;
    /** @type {?} */
    GroupFooterRowStylerDirective.prototype.dg;
}
