// @gsp-wf/wf-process-editor/esm2015/lib/views/bill-type-help/entity/DocuTypeHelperEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
export class DocuTypeHelperEntity extends JsonConvertableImpl {
    constructor() {
        super();
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.id = null;
        this.moduleName = null;
        this.bizKindName = null;
        this.bizTypeName = null;
        this.code = null;
        this.name = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "id", this.id);
        BpmnModelJsonHelper.SetAttributeValue(obj, "moduleName", this.moduleName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizKindName", this.bizKindName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeName", this.bizTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.id = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "id");
        this.moduleName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "moduleName");
        this.bizKindName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizKindName");
        this.bizTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeName");
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
    }
}
if (false) {
    /** @type {?} */
    DocuTypeHelperEntity.prototype.id;
    /** @type {?} */
    DocuTypeHelperEntity.prototype.moduleName;
    /** @type {?} */
    DocuTypeHelperEntity.prototype.bizKindName;
    /** @type {?} */
    DocuTypeHelperEntity.prototype.bizTypeName;
    /** @type {?} */
    DocuTypeHelperEntity.prototype.code;
    /** @type {?} */
    DocuTypeHelperEntity.prototype.name;
}
