// @gsp-wf/wf-task-handler/esm2015/lib/wf-task-handler.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, ComponentFactoryResolver, ApplicationRef } from '@angular/core';
import { WfApprovalLogsService } from '@gsp-wf/wf-approval-logs';
import { ForecastProcessPayload, WFFlowchartService } from '@gsp-wf/ui-flowchart';
import { BsModalService } from '@farris/ui-modal';
import { Subject, of } from 'rxjs';
import { TaskService } from './service/task.service';
import { TaskAction } from './entity/task-action';
import { NotifyService } from '@farris/ui-notify';
import { I18nService } from './service/i18n/i18n.service';
import { FrameworkService } from '@gsp-sys/rtf-common';
import { TaskEntity } from './entity/task-entity';
import { TaskExecuteComponent } from './components/task-execute/task-execute.component';
import { switchMap } from 'rxjs/operators';
import { StartProcessPayload, CancelSubmitPayload } from './entity/start-process-payload';
import { ProcessSelectorComponent } from './submit/process-selector/process-selector.component';
import { LoadingService } from '@farris/ui-loading';
import { MessagerService } from '@farris/ui-messager';
import { ConfigCode, OwnerType } from './entity/task-center-config';
import { CommentManageComponent } from './components/comment-manage/comment-manage.component';
export class WfTaskHandlerService {
    /**
     * @param {?} logService
     * @param {?} flowchartService
     * @param {?} injector
     * @param {?} resolver
     * @param {?} modalService
     * @param {?} taskService
     * @param {?} notify
     * @param {?} i18nService
     * @param {?} frameworkService
     * @param {?} appRef
     */
    constructor(logService, flowchartService, injector, resolver, modalService, taskService, notify, i18nService, frameworkService, appRef) {
        this.logService = logService;
        this.flowchartService = flowchartService;
        this.injector = injector;
        this.resolver = resolver;
        this.modalService = modalService;
        this.taskService = taskService;
        this.notify = notify;
        this.i18nService = i18nService;
        this.frameworkService = frameworkService;
        this.appRef = appRef;
        this.subject = new Subject();
        this.curLanguage = localStorage.getItem('languageCode') || 'zh-CHS'; //系统当前语言
        //系统当前语言
        this.batchSubmitNum = {
            //批量提交数量
            successNum: 0,
            failedNum: 0,
            errorNum: 0,
        };
        this.batchCancelNum = {
            //批量取消提交数量
            successNum: 0,
            errorNum: 0
        };
        this.loading = this.injector.get(LoadingService);
        this.message = this.injector.get(MessagerService);
    }
    /**
     * 处理单条任务
     * @param {?} task
     * @param {?} action 任务动作
     * @param {?=} otherParams
     * @return {?}
     */
    execute(task, action, otherParams) {
        switch (action.code) {
            case 'ViewBizLog': //审批日志。如果是inSuite，根据流程实例打开
                this.logService.showByProcInstId(task.processInstanceId);
                break;
            case 'ViewFlowChart': //查看流程
                this.flowchartService.viewFlowchartByProcInstIdAndDataId(task.processInstanceId, task.refBizId);
                break;
            case 'ViewExecuteForm': // 联查单据
                this.openApp(task, action, otherParams);
                break;
            default: //办理类任务
                this.complete(task, action);
        }
        return this.subject.asObservable();
    }
    /**
     * @param {?} assigneeInfo
     * @param {?} openFromIframe
     * @return {?}
     */
    assignByInfo(assigneeInfo, openFromIframe) {
        /** @type {?} */
        const assignSubject = new Subject();
        return assignSubject.asObservable();
    }
    /**
     * @private
     * @param {?} task
     * @param {?} action
     * @return {?}
     */
    complete(task, action) {
        /** @type {?} */
        const compFactory = this.resolver.resolveComponentFactory(TaskExecuteComponent);
        /** @type {?} */
        const compRef = compFactory.create(this.injector);
        compRef.instance.setTask(task);
        compRef.instance.setAction(action);
        compRef.instance.addPadding();
        /** @type {?} */
        const modalConfig = {
            title: action.name ? action.name : this.i18nService.getValue('task.title'),
            width: 550,
            height: compRef.instance.dialogHeight,
            showButtons: true,
            buttons: compRef.instance.buttons
        };
        /** @type {?} */
        const dialog = this.modalService.show(compRef, modalConfig);
        compRef.instance.getSubject().subscribe((/**
         * @param {?} re
         * @return {?}
         */
        re => {
            re && re.result && this.subject.next(true);
            re && re.result && dialog.close();
        }));
    }
    /**
     * 打开审批联查界面
     * @param {?} entity 任务实体
     * @param {?} action 动作实体
     * @param {?=} otherParams
     * @return {?}
     */
    openApp(entity, action, otherParams) {
        /** @type {?} */
        const sourceId = entity.sourceId;
        /** @type {?} */
        const formInfoLoading = this.loading.show();
        this.taskService.getTaskFormInfo(sourceId).subscribe((/**
         * @param {?} formInfo
         * @return {?}
         */
        (formInfo) => {
            formInfoLoading.close();
            if (!formInfo) {
                this.notify.info(this.i18nService.getValue('task.notFoundBill'));
                return;
            }
            /** @type {?} */
            const options = {
                appType: 'menu',
                funcId: '',
                appId: '',
                appEntrance: '',
                tabId: sourceId,
                isNewTab: true
            };
            // 传工作流上下文参数
            /** @type {?} */
            const parameters = new Map();
            // 功能菜单情况
            if (formInfo.appId) {
                // 默认联查审批界面
                if (formInfo.formType === '1') { //表单嵌流程
                    parameters.set('workItemId', sourceId);
                    parameters.set('procInstId', action.parameters[0].value);
                    parameters.set('taskState', entity.state);
                    parameters.set('isAssign', entity.actions.some((/**
                     * @param {?} action
                     * @return {?}
                     */
                    action => action.code === 'AssignUser')));
                    formInfo.parameters.forEach((/**
                     * @param {?} item
                     * @return {?}
                     */
                    (item) => parameters.set(item.code, item.value)));
                    options.funcId = formInfo.appId;
                }
                else { //流程嵌表单
                    options.funcId = 'WF0120';
                    parameters.set('task', entity);
                    parameters.set('formInfo', formInfo);
                    otherParams && parameters.set('taskFilter', otherParams.taskFilter);
                }
            }
            else if (formInfo.url) { //表单格式
                options.funcId = 'WF0120';
                parameters.set('task', entity);
                parameters.set('formInfo', formInfo);
                otherParams && parameters.set('taskFilter', otherParams.taskFilter);
            }
            else {
                this.notify.info(this.i18nService.getValue('task.notFoundBill'));
                return;
            }
            options.entityParams = parameters;
            options.queryStringParams = parameters; //兼容IDP表单
            this.openMenu(options, this.i18nService.getValue('static.taskForm'));
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            formInfoLoading.close();
            this.taskService.error(error);
        }));
    }
    /**
     * 打开菜单，区分是否inSuite环境，用不同方式打开菜单
     * @private
     * @param {?} options 打开参数
     * @param {?} menuTitle 菜单名称
     * @return {?}
     */
    openMenu(options, menuTitle) {
        if (this.taskService.isInSuite()) {
            options.menuTitle = menuTitle;
            /** @type {?} */
            var extObj = ((/** @type {?} */ (window))).gspframeworkService.rtf.extendMethod.getExtObj();
            extObj.iGIXMenuOpen(options);
        }
        else {
            /** @type {?} */
            const openMenuLoading = this.loading.show();
            this.frameworkService.openMenu$(options).subscribe((/**
             * @param {?} re
             * @return {?}
             */
            re => {
                openMenuLoading.close();
            }), (/**
             * @param {?} error
             * @return {?}
             */
            error => {
                openMenuLoading.close();
                this.taskService.error(error);
            }));
        }
    }
    /**
     * 返回动作列表，表单嵌流程用到
     * @param {?} workItemId
     * @return {?}
     */
    refreshActions(workItemId) {
        return this.taskService.getWorkItemActions(workItemId).pipe(switchMap((/**
         * @param {?} actionList
         * @return {?}
         */
        (actionList) => {
            return of(actionList.filter((/**
             * @param {?} action
             * @return {?}
             */
            action => action.code !== 'ViewBizLog' &&
                action.code !== 'ViewFlowChart' &&
                action.code !== 'ViewExecuteForm')));
        })));
    }
    /**
     * 根据流程实例ID和动作Code办理任务（暂未用到 by fangpengfei\@2020.06.22）
     * @param {?} procInstId 流程实例ID
     * @param {?} state 任务状态
     * @param {?} actionCode 动作Code
     * @return {?}
     */
    executeByProcInstId(procInstId, state, actionCode) {
        /** @type {?} */
        var action = new TaskAction();
        action.code = actionCode;
        /** @type {?} */
        const subject = new Subject();
        this.taskService.getTaskByProcInstId(procInstId, state).subscribe((/**
         * @param {?} task
         * @return {?}
         */
        (task) => {
            if (task) {
                this.execute(task, action).subscribe((/**
                 * @param {?} re
                 * @return {?}
                 */
                re => {
                    subject.next(re);
                }));
            }
            else {
                this.notify.info(this.i18nService.getValue('task.noTask'));
            }
        }));
        return subject.asObservable();
    }
    /**
     * 根据workItemId 办理任务
     * @param {?} workItemId 工作项ID
     * @param {?} action 动作实体
     * @return {?}
     */
    executeByWorkItemId(workItemId, action) {
        /** @type {?} */
        let task = new TaskEntity();
        task.sourceId = workItemId;
        return this.execute(task, action);
    }
    /**
     * 提交审批
     * @param {?} payload 上下文参数
     * @return {?} subject true：提交成功，false：提交失败
     */
    startProcess(payload) {
        /** @type {?} */
        const subject = new Subject();
        /** @type {?} */
        const configLoad = this.loading.show();
        this.taskService.getTaskCenterConfigByCode(ConfigCode.If_Show_Submit_Dialog, OwnerType.BizDefKey, payload.bizDefKey).subscribe((/**
         * @param {?} re
         * @return {?}
         */
        re => {
            configLoad.close();
            if (re && re.currentValue === 'false') { //自定义无需弹框
                //自定义无需弹框
                /** @type {?} */
                const simulateLoad = this.loading.show();
                this.taskService.startProcessSimulate(payload).subscribe((/**
                 * @param {?} response
                 * @return {?}
                 */
                (response) => {
                    simulateLoad.close();
                    if (!response || response.needInteraction === false) { //预解析无需交互
                        //预解析无需交互
                        /** @type {?} */
                        const executeLoad = this.loading.show();
                        this.taskService.startProcess(payload).subscribe((/**
                         * @param {?} response
                         * @return {?}
                         */
                        (response) => {
                            executeLoad.close();
                            if (payload.withResponse) {
                                subject.next(response);
                            }
                            else {
                                subject.next(true);
                            }
                            this.handleSubmitMessage(ConfigCode.Info_After_Submit, payload.bizDefKey);
                        }), (/**
                         * @param {?} error
                         * @return {?}
                         */
                        error => {
                            executeLoad.close();
                            if (payload.withResponse) {
                                subject.next(null);
                            }
                            else {
                                subject.next(false);
                            }
                            this.taskService.error(error);
                        }));
                    }
                    else { //预解析需要交互
                        this.showStartProcessDialog(payload).subscribe((/**
                         * @param {?} re
                         * @return {?}
                         */
                        re => {
                            subject.next(re);
                        }));
                    }
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                error => {
                    simulateLoad.close();
                    subject.next(null);
                    this.taskService.error(error);
                }));
            }
            else { //无是否弹框配置
                this.showStartProcessDialog(payload).subscribe((/**
                 * @param {?} re
                 * @return {?}
                 */
                re => {
                    subject.next(re);
                }));
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            configLoad.close();
            this.taskService.error(error);
        }));
        return subject.asObservable();
    }
    /**
     * 提交审批弹框
     * @private
     * @param {?} payload 参数
     * @return {?}
     */
    showStartProcessDialog(payload) {
        /** @type {?} */
        const subject = new Subject();
        /** @type {?} */
        const compFactory = this.resolver.resolveComponentFactory(ProcessSelectorComponent);
        /** @type {?} */
        const compRef = compFactory.create(this.injector);
        compRef.instance.setParam(payload).subscribe((/**
         * @param {?} re
         * @return {?}
         */
        (re) => {
            /** @type {?} */
            const modalConfig = {
                title: this.i18nService.getValue('startProcess.title'),
                width: 520,
                height: re && re.dialogHeight,
                showButtons: true,
                buttons: compRef.instance.buttons,
                opened: (/**
                 * @return {?}
                 */
                () => {
                    setTimeout((/**
                     * @return {?}
                     */
                    () => {
                        compRef.changeDetectorRef && compRef.changeDetectorRef.detectChanges();
                    }), 100);
                })
            };
            /** @type {?} */
            const dialog = this.modalService.show(compRef, modalConfig);
            if (payload.withResponse === true) { //内部使用
                compRef.instance.getSubjectWithResponse().subscribe((/**
                 * @param {?} re
                 * @return {?}
                 */
                re => {
                    subject.next(re);
                    dialog.close();
                }));
                compRef.instance.getSubjectCancel().subscribe((/**
                 * @return {?}
                 */
                () => {
                    dialog.close();
                }));
            }
            else { //已有业务组调用
                compRef.instance.getSubject().subscribe((/**
                 * @param {?} re
                 * @return {?}
                 */
                re => {
                    subject.next(re);
                    dialog.close();
                }));
                compRef.instance.getSubjectCancel().subscribe((/**
                 * @param {?} re
                 * @return {?}
                 */
                re => {
                    subject.next(re);
                    dialog.close();
                }));
            }
        }));
        return subject.asObservable();
    }
    /**
     * 取消提交
     * \@payload payload 参数
     * @param {?} payload
     * @return {?}
     */
    cancelSubmit(payload) {
        /** @type {?} */
        const subject = new Subject();
        /** @type {?} */
        const load = this.loading.show();
        this.taskService.cancelSubmit(payload).subscribe((/**
         * @param {?} re
         * @return {?}
         */
        (re) => {
            load.close();
            subject.next(true);
            if (re && re.needInteraction) {
                this.notify.info(this.i18nService.getValue('startProcess.cancelSubmitInfo'));
            }
            else {
                this.handleSubmitMessage(ConfigCode.Info_After_Cancel_Submit, payload.bizDefKey);
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            load.close();
            subject.error(false);
            this.taskService.error(error);
        }));
        return subject.asObservable();
    }
    /**
     * 提交审批
     * @param {?} dataId 表单id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    submitWithBizDefKey(dataId, bizDefKey) {
        if (!dataId) {
            this.notify.warning(this.i18nService.getValue('startProcess.dataIdIsNull'));
            return;
        }
        if (!bizDefKey) {
            this.message.error(this.i18nService.getValue('startProcess.bizDefKeyIsNull'));
            return;
        }
        /** @type {?} */
        var payload = new StartProcessPayload();
        payload.dataId = dataId;
        payload.bizDefKey = bizDefKey;
        payload.withResponse = true;
        return this.startProcess(payload);
    }
    /**
     * 取消提交审批
     * @param {?} dataId 表单id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    cancelSubmitWithDataId(dataId, bizDefKey) {
        if (!dataId) {
            this.notify.warning(this.i18nService.getValue('startProcess.dataIdIsNull'));
            return;
        }
        if (!bizDefKey) {
            this.message.error(this.i18nService.getValue('startProcess.bizDefKeyIsNull'));
            return;
        }
        /** @type {?} */
        var cancelSubmitPayload = new CancelSubmitPayload();
        cancelSubmitPayload.dataId = dataId;
        cancelSubmitPayload.bizDefKey = bizDefKey;
        return this.cancelSubmit(cancelSubmitPayload);
    }
    /**
     * 批量提交审批
     * @param {?} dataIds 表单ID集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    batchSubmitWithBizDefKey(dataIds, bizDefKey) {
        if (!dataIds || !dataIds.length) {
            this.notify.warning(this.i18nService.getValue('startProcess.dataIdIsEmpty'));
            return;
        }
        if (!bizDefKey) {
            this.message.error(this.i18nService.getValue('startProcess.bizDefKeyIsNull'));
            return;
        }
        if (dataIds.length === 1) {
            /** @type {?} */
            var startPayload = new StartProcessPayload();
            startPayload.bizDefKey = bizDefKey;
            startPayload.dataId = dataIds[0];
            return this.startProcess(startPayload);
        }
        /** @type {?} */
        const subject = new Subject();
        /** @type {?} */
        const load = this.loading.show();
        this.batchSubmitNum.successNum = 0;
        this.batchSubmitNum.failedNum = 0;
        this.batchSubmitNum.errorNum = 0;
        this.batchSubmitHelpFunction([...dataIds], bizDefKey, (/**
         * @return {?}
         */
        () => {
            load.close();
            subject.next(true);
        }));
        return subject.asObservable();
    }
    /**
     * 批量提交审批辅助函数
     * @private
     * @param {?} dataIds
     * @param {?} bizDefKey
     * @param {?} handler
     * @return {?}
     */
    batchSubmitHelpFunction(dataIds, bizDefKey, handler) {
        if (dataIds && dataIds.length > 0) {
            /** @type {?} */
            const dataId = dataIds.shift();
            //引用类型转换
            /** @type {?} */
            let startPayload = new StartProcessPayload();
            startPayload.bizDefKey = bizDefKey;
            startPayload.dataId = dataId;
            startPayload.variables = { simulate: true };
            this.taskService.startProcess(startPayload).subscribe((/**
             * @param {?} response
             * @return {?}
             */
            (response) => {
                if (response.needInteraction) { //需要交互
                    this.batchSubmitNum.failedNum++;
                    this.batchSubmitHelpFunction(dataIds, bizDefKey, handler);
                }
                else { //不需要交互
                    startPayload.variables = { simulate: false };
                    this.taskService.startProcess(startPayload).subscribe((/**
                     * @param {?} response
                     * @return {?}
                     */
                    (response) => {
                        this.batchSubmitNum.successNum++;
                        this.batchSubmitHelpFunction(dataIds, bizDefKey, handler);
                    }), (/**
                     * @return {?}
                     */
                    () => {
                        this.batchSubmitNum.errorNum++;
                        this.batchSubmitHelpFunction(dataIds, bizDefKey, handler);
                    }));
                }
            }), (/**
             * @return {?}
             */
            () => {
                this.batchSubmitNum.errorNum++;
                this.batchSubmitHelpFunction(dataIds, bizDefKey, handler);
            }));
        }
        else { //批量执行完毕
            this.showBatchSubmitInfo(this.batchSubmitNum, bizDefKey);
            handler(); //执行回调函数
        }
    }
    /**
     * 批量取消提交审批
     * @param {?} dataIds 表单ID集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    batchCancelSubmitWithDataId(dataIds, bizDefKey) {
        if (!dataIds || !dataIds.length) {
            this.notify.warning(this.i18nService.getValue('startProcess.dataIdIsEmpty'));
            return;
        }
        if (!bizDefKey) {
            this.message.error(this.i18nService.getValue('startProcess.bizDefKeyIsNull'));
            return;
        }
        if (dataIds.length === 1) {
            return this.cancelSubmitWithDataId(dataIds[0], bizDefKey);
        }
        /** @type {?} */
        const subject = new Subject();
        /** @type {?} */
        const load = this.loading.show();
        this.batchCancelNum.successNum = 0;
        this.batchCancelNum.errorNum = 0;
        this.batchCancelHelpFunction([...dataIds], bizDefKey, (/**
         * @return {?}
         */
        () => {
            load.close();
            subject.next(true);
        }));
        return subject.asObservable();
    }
    /**
     * 批量取消辅助函数
     * @private
     * @param {?} dataIds
     * @param {?} bizDefKey
     * @param {?} handler
     * @return {?}
     */
    batchCancelHelpFunction(dataIds, bizDefKey, handler) {
        if (dataIds && dataIds.length) {
            /** @type {?} */
            const dataId = dataIds.shift();
            //引用类型转换
            /** @type {?} */
            let cancelPayload = new CancelSubmitPayload();
            cancelPayload.dataId = dataId;
            cancelPayload.bizDefKey = bizDefKey;
            this.taskService.cancelSubmit(cancelPayload).subscribe((/**
             * @return {?}
             */
            () => {
                this.batchCancelNum.successNum++;
                this.batchCancelHelpFunction(dataIds, bizDefKey, handler);
            }), (/**
             * @return {?}
             */
            () => {
                this.batchCancelNum.errorNum++;
                this.batchCancelHelpFunction(dataIds, bizDefKey, handler);
            }));
        }
        else {
            this.showBatchCancelSubmitInfo(this.batchCancelNum, bizDefKey);
            handler();
        }
    }
    /**
     * 子表提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataId 子表id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    childSubmit(parentDataId, childDataId, bizDefKey) {
        if (!parentDataId || !childDataId) {
            this.notify.warning(this.i18nService.getValue('startProcess.dataIdIsNull'));
            return;
        }
        if (!bizDefKey) {
            this.message.error(this.i18nService.getValue('startProcess.bizDefKeyIsNull'));
            return;
        }
        /** @type {?} */
        var startProcessPayload = new StartProcessPayload();
        startProcessPayload.dataId = `${parentDataId},${childDataId}`;
        startProcessPayload.bizDefKey = bizDefKey;
        return this.startProcess(startProcessPayload);
    }
    /**
     * 子表取消提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataId 子表id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    childCancelSubmit(parentDataId, childDataId, bizDefKey) {
        if (!parentDataId || !childDataId) {
            this.notify.warning(this.i18nService.getValue('startProcess.dataIdIsNull'));
            return;
        }
        if (!bizDefKey) {
            this.message.error(this.i18nService.getValue('startProcess.bizDefKeyIsNull'));
            return;
        }
        /** @type {?} */
        var cancelSubmitPayload = new CancelSubmitPayload();
        cancelSubmitPayload.dataId = `${parentDataId},${childDataId}`;
        cancelSubmitPayload.bizDefKey = bizDefKey;
        return this.cancelSubmit(cancelSubmitPayload);
    }
    /**
     * 子表批量提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataIds 子表id集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    childBatchSubmit(parentDataId, childDataIds, bizDefKey) {
        if (!parentDataId || !childDataIds) {
            this.notify.warning(this.i18nService.getValue('startProcess.dataIdIsNull'));
            return;
        }
        if (!bizDefKey) {
            this.message.error(this.i18nService.getValue('startProcess.bizDefKeyIsNull'));
            return;
        }
        /** @type {?} */
        const dataIds = childDataIds.map((/**
         * @param {?} childDataId
         * @return {?}
         */
        childDataId => `${parentDataId},${childDataId}`));
        return this.batchSubmitWithBizDefKey(dataIds, bizDefKey);
    }
    /**
     * 子表批量取消提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataIds 子表id集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    childBatchCancelSubmit(parentDataId, childDataIds, bizDefKey) {
        if (!parentDataId || !childDataIds) {
            this.notify.warning(this.i18nService.getValue('startProcess.dataIdIsNull'));
            return;
        }
        if (!bizDefKey) {
            this.message.error(this.i18nService.getValue('startProcess.bizDefKeyIsNull'));
            return;
        }
        /** @type {?} */
        const dataIds = childDataIds.map((/**
         * @param {?} childDataId
         * @return {?}
         */
        childDataId => `${parentDataId},${childDataId}`));
        return this.batchCancelSubmitWithDataId(dataIds, bizDefKey);
    }
    /**
     * 批量提交提示信息
     * @private
     * @param {?} num
     * @param {?} bizDefKey
     * @return {?}
     */
    showBatchSubmitInfo(num, bizDefKey) {
        if (num.successNum && !num.failedNum && !num.errorNum) { //批量全部成功
            this.taskService.getTaskCenterConfigByCode(ConfigCode.Info_After_Submit, OwnerType.BizDefKey, bizDefKey).subscribe((/**
             * @param {?} config
             * @return {?}
             */
            (config) => {
                if (this.curLanguage === 'en') {
                    this.notify.success(JSON.parse(config.currentValue).en);
                }
                if (this.curLanguage === 'zh-CHS') {
                    this.notify.success(JSON.parse(config.currentValue).zh_CHS);
                }
            }));
        }
        else { //非全部成功
            //非全部成功
            /** @type {?} */
            let message = '';
            num.successNum && (message += this.i18nService.getValue('batchStartProcess.success') + `【${num.successNum}】` + this.i18nService.getValue('batchStartProcess.row'));
            num.errorNum && (message += this.i18nService.getValue('batchStartProcess.error') + `【${num.errorNum}】` + this.i18nService.getValue('batchStartProcess.row'));
            num.failedNum && (message += this.i18nService.getValue('batchStartProcess.failed') + `【${num.failedNum}】` + this.i18nService.getValue('batchStartProcess.row'));
            message += this.i18nService.getValue('batchStartProcess.single');
            this.message.show('warning', message, { safeHtml: false });
        }
    }
    /**
     * 批量取消提交提示信息
     * @private
     * @param {?} num
     * @param {?} bizDefKey
     * @return {?}
     */
    showBatchCancelSubmitInfo(num, bizDefKey) {
        if (num.successNum && !num.errorNum) { //批量全部成功
            this.taskService.getTaskCenterConfigByCode(ConfigCode.Info_After_Cancel_Submit, OwnerType.BizDefKey, bizDefKey).subscribe((/**
             * @param {?} config
             * @return {?}
             */
            (config) => {
                if (this.curLanguage === 'en') {
                    this.notify.success(JSON.parse(config.currentValue).en);
                }
                if (this.curLanguage === 'zh-CHS') {
                    this.notify.success(JSON.parse(config.currentValue).zh_CHS);
                }
            }));
        }
        else {
            /** @type {?} */
            let message = '';
            num.successNum && (message += this.i18nService.getValue('startProcess.success') + `【${num.successNum}】` + this.i18nService.getValue('startProcess.row'));
            num.errorNum && (message += this.i18nService.getValue('startProcess.failed') + `【${num.errorNum}】` + this.i18nService.getValue('startProcess.row'));
            this.message.show('warning', message, { safeHtml: false });
        }
    }
    /**
     * @param {?} code
     * @param {?} ownerId
     * @return {?}
     */
    handleSubmitMessage(code, ownerId) {
        /** @type {?} */
        const load = this.loading.show();
        this.taskService.getTaskCenterConfigByCode(code, OwnerType.BizDefKey, ownerId).subscribe((/**
         * @param {?} re
         * @return {?}
         */
        re => {
            load.close();
            if (re && re.currentValue) {
                /** @type {?} */
                let message = '';
                /** @type {?} */
                const currentValue = JSON.parse(re.currentValue);
                this.curLanguage === 'zh-CHS' && (message = currentValue.zh_CHS);
                this.curLanguage === 'en' && (message = currentValue.en);
                this.curLanguage === 'zh-CHT' && (message = currentValue.zh_CHT);
                this.notify.success(message);
            }
            else {
                code === ConfigCode.Info_After_Submit && this.notify.success(this.i18nService.getValue('startProcess.submitSuccess'));
                code === ConfigCode.Info_After_Cancel_Submit && this.notify.success(this.i18nService.getValue('startProcess.cancelSubmitSuccess'));
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            load.close();
            this.taskService.error(error);
        }));
    }
    /**
     * 表单上触发流程预解析操作
     * 已封装为表单的命令操作
     * @return {?}
     */
    processSimulate() {
        /** @type {?} */
        const subject = new Subject();
        /** @type {?} */
        const gspWfEventService = ((/** @type {?} */ (window))).parent.gspWfEventService;
        gspWfEventService.simulate((/**
         * @return {?}
         */
        () => {
            subject.next(true);
        }));
        return subject.asObservable();
    }
    /**
     * 根据单据id查看流程图
     * 已封装为表单命令和操作
     * @param {?} dataId
     * @return {?}
     */
    viewFlowChartByDataId(dataId) {
        if (!dataId) {
            this.notify.warning(this.i18nService.getValue('startProcess.dataIdIsNull'));
            return;
        }
        this.flowchartService.viewFlowChartByDataId(dataId);
    }
    /**
     * 查看流程图（新，支持预览）
     * 已封装为表单命令和操作
     * @param {?} bizDefKey 流程分类id
     * @param {?} dataId 表单数据id
     * @return {?}
     */
    viewFlowChartWithForecast(bizDefKey, dataId) {
        if (!dataId) {
            this.notify.warning(this.i18nService.getValue('startProcess.dataIdIsNull'));
            return;
        }
        if (!bizDefKey) {
            this.message.error(this.i18nService.getValue('startProcess.bizDefKeyIsNull'));
            return;
        }
        /** @type {?} */
        let payload = new ForecastProcessPayload();
        payload.bizDefKey = bizDefKey;
        payload.dataId = dataId;
        this.flowchartService.viewProcess(payload);
    }
    /**
     * 查看审批记录（弹框）
     * 已封装为表单命令
     * @param {?} dataId 表单数据id
     * @param {?} style 其他参数
     * @return {?}
     */
    viewApprovalLog(dataId, style) {
        if (!dataId) {
            this.notify.warning(this.i18nService.getValue('startProcess.dataIdIsNull'));
            return;
        }
        /** @type {?} */
        const parameters = {
            dataId: dataId,
            style: style,
        };
        this.logService.showApprovalLog(parameters);
    }
    /**
     * 获取用户的常用意见（已暴露给farris）
     * @param {?} filter
     * @return {?} 用户常用意见list
     */
    getCommonComments(filter) {
        return this.taskService.getCommentList(filter);
    }
    /**
     * 展示常用意见管理dialog（已暴露给farris）
     * @param {?} filter
     * @return {?}
     */
    showCommentManageDialog(filter) {
        /** @type {?} */
        let subject = new Subject();
        /** @type {?} */
        const compFactory = this.resolver.resolveComponentFactory(CommentManageComponent);
        /** @type {?} */
        const compRef = compFactory.create(this.injector);
        /** @type {?} */
        const options = {
            title: this.i18nService.getValue('comment.static.title'),
            width: 740,
            height: 440,
            showButtons: true,
            buttons: compRef.instance.buttons
        };
        compRef.instance.filter = filter;
        /** @type {?} */
        const dialog = this.modalService.show(compRef, options);
        compRef.instance.getConfirmSUbject().subscribe((/**
         * @param {?} re
         * @return {?}
         */
        re => {
            dialog.close();
            subject.next(re);
        }));
        compRef.instance.getCancelSubject().subscribe((/**
         * @param {?} re
         * @return {?}
         */
        re => {
            dialog.close();
            subject.next(re);
        }));
        return subject.asObservable();
    }
}
WfTaskHandlerService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
WfTaskHandlerService.ctorParameters = () => [
    { type: WfApprovalLogsService },
    { type: WFFlowchartService },
    { type: Injector },
    { type: ComponentFactoryResolver },
    { type: BsModalService },
    { type: TaskService },
    { type: NotifyService },
    { type: I18nService },
    { type: FrameworkService },
    { type: ApplicationRef }
];
if (false) {
    /** @type {?} */
    WfTaskHandlerService.prototype.subject;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.loading;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.message;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.curLanguage;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.batchSubmitNum;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.batchCancelNum;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.logService;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.flowchartService;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.taskService;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.i18nService;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.frameworkService;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.appRef;
}
