// @farris/external-integration/esm5/lib/component/external-integration/external-integration-service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component/external-integration/external-integration-service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from "@angular/common/http";
import { Injectable, Output, EventEmitter } from "@angular/core";
import { VariableParseService, ViewModel } from "@farris/devkit";
var ExternalIntegrationService = /** @class */ (function () {
    function ExternalIntegrationService(http, variableParseService, vm) {
        this.http = http;
        this.variableParseService = variableParseService;
        this.vm = vm;
        this.sapData = new EventEmitter();
        this.sapUrl = "/api/runtime/icc/v1.0/externalServiceOpenApi/invokeService";
    }
    /**
     * 调用外部服务
     */
    /**
     * 调用外部服务
     * @param {?} externalParams
     * @param {?} serviceCode
     * @return {?}
     */
    ExternalIntegrationService.prototype.queryExtIntegrationService = /**
     * 调用外部服务
     * @param {?} externalParams
     * @param {?} serviceCode
     * @return {?}
     */
    function (externalParams, serviceCode) {
        var _this = this;
        /** @type {?} */
        var req = this.assembleSapRequest(externalParams, serviceCode);
        this.http.post(this.sapUrl, req).subscribe((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            /** @type {?} */
            var respObj = Object.create(response);
            if (!respObj || !respObj.success) {
                return;
            }
            /** @type {?} */
            var resp = Object.create(respObj);
            /** @type {?} */
            var sapData = resp.data;
            _this.sapData.emit(sapData);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            console.log(error);
            _this.sapData.emit({});
        }));
    };
    /**
     * 组装sap api入参
     */
    /**
     * 组装sap api入参
     * @private
     * @param {?} externalParams
     * @param {?} serviceCode
     * @return {?}
     */
    ExternalIntegrationService.prototype.assembleSapRequest = /**
     * 组装sap api入参
     * @private
     * @param {?} externalParams
     * @param {?} serviceCode
     * @return {?}
     */
    function (externalParams, serviceCode) {
        /** @type {?} */
        var params = JSON.parse(JSON.stringify(externalParams));
        /** @type {?} */
        var requestParam = {
            code: serviceCode,
            params: this.variableParseService.parse(params, this.vm.frameContext),
        };
        return requestParam;
    };
    ExternalIntegrationService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ExternalIntegrationService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: VariableParseService },
        { type: ViewModel }
    ]; };
    ExternalIntegrationService.propDecorators = {
        sapData: [{ type: Output }]
    };
    return ExternalIntegrationService;
}());
export { ExternalIntegrationService };
if (false) {
    /** @type {?} */
    ExternalIntegrationService.prototype.sapData;
    /** @type {?} */
    ExternalIntegrationService.prototype.sapUrl;
    /**
     * @type {?}
     * @private
     */
    ExternalIntegrationService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    ExternalIntegrationService.prototype.variableParseService;
    /**
     * @type {?}
     * @private
     */
    ExternalIntegrationService.prototype.vm;
}
