// @farris/external-integration/esm5/lib/component/external-integration/grid-external-integration.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component/external-integration/grid-external-integration.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, Injector, Renderer2, ViewChild, } from "@angular/core";
import { RuntimeStateService } from "@farris/ui-common";
import { GRID_EDITORS } from "@farris/ui-datagrid";
import { DatagridBaseEditorDirective } from "@farris/ui-datagrid-editors";
import { ExternalIntegrationComponent } from "./external-integration.component";
/** @type {?} */
var ExtIntgrtnDefautOption = {
    readonly: false,
    // 是否只读
    placeholder: "请选择",
    viewType: "tag"
};
var DatagridExtIntgrtnComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridExtIntgrtnComponent, _super);
    function DatagridExtIntgrtnComponent(render, el, rts, injector) {
        var _this = _super.call(this, render, el, injector) || this;
        _this.rts = rts;
        _this.stopPropagation = false;
        return _this;
    }
    /**
     * @return {?}
     */
    DatagridExtIntgrtnComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        _super.prototype.ngOnInit.call(this);
        this.inputElement = this.instance.input.nativeElement;
        this.options = Object.assign({}, ExtIntgrtnDefautOption, this.options);
        this.column.editor.options.onBlur = (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            e.editorRef.startPending();
            e.editorRef.formControl.setValue(_this.instance.fieldText);
            _this.instance.callSapService();
            _this.instance.afterOnblur.subscribe((/**
             * @return {?}
             */
            function () {
                e.editorRef.endPending();
            }));
        });
    };
    /**
     * @return {?}
     */
    DatagridExtIntgrtnComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngAfterViewInit.call(this);
    };
    DatagridExtIntgrtnComponent.decorators = [
        { type: Component, args: [{
                    selector: "grid-external-integration",
                    template: "\n    <div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup farris-group-auto\">\n      <web-external-integration\n        #extIntgrtn\n        style=\"width: 100%\"\n        extIntegration-dataMapping\n        [placeholder]=\"options.placeholder\"\n        [readonly]=\"options.readonly\"\n        [mapFields]=\"options.mapFields\"\n        [serviceCode] = \"options.serviceCode\"\n        [externalParams] = \"options.externalParams\"\n        [extTableSchemas] = \"options.extTableSchemas\"\n        [formControlName]=\"column.field\"\n        [mappingType] = \"'grid'\"\n        \n      ></web-external-integration>\n    </div>\n  "
                }] }
    ];
    /** @nocollapse */
    DatagridExtIntgrtnComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: RuntimeStateService },
        { type: Injector }
    ]; };
    DatagridExtIntgrtnComponent.propDecorators = {
        instance: [{ type: ViewChild, args: ["extIntgrtn",] }]
    };
    return DatagridExtIntgrtnComponent;
}(DatagridBaseEditorDirective));
export { DatagridExtIntgrtnComponent };
if (false) {
    /** @type {?} */
    DatagridExtIntgrtnComponent.prototype.instance;
    /** @type {?} */
    DatagridExtIntgrtnComponent.prototype.stopPropagation;
    /**
     * @type {?}
     * @private
     */
    DatagridExtIntgrtnComponent.prototype.rts;
}
/** @type {?} */
export var ExtIntgrtnDataGridEditorProvider = {
    provide: GRID_EDITORS,
    useValue: { name: "external-integration", value: DatagridExtIntgrtnComponent },
    multi: true
};
