// @farris/lookup-config/esm5/lib/lookup-config.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/lookup-config.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ComponentFactoryResolver, Injector } from '@angular/core';
import { LookupConfigComponent } from './lookup-config.component';
import { BsModalService } from '@farris/ui-modal';
import { Subject } from 'rxjs';
import { NotifyService } from '@farris/ui-notify';
import { ConfigService } from './config.service';
import { HttpClient } from '@angular/common/http';
import { LocaleService } from '@farris/ui-locale';
import * as i0 from "@angular/core";
import * as i1 from "@farris/ui-modal";
import * as i2 from "@farris/ui-notify";
import * as i3 from "./config.service";
import * as i4 from "@angular/common/http";
var LookupConfigService = /** @class */ (function () {
    function LookupConfigService(cfr, injector, modalService, notifyService, configService, http) {
        this.cfr = cfr;
        this.injector = injector;
        this.modalService = modalService;
        this.notifyService = notifyService;
        this.configService = configService;
        this.http = http;
        this.lookupconfigData = new Subject();
        this.lookupconfigDatas = new Subject();
        this.lookupTitle = '帮助配置';
        this.modalSure = '确定';
        this.modalCancel = '取消';
        this.successSave = '保存成功';
        this.helpIdError = '请选择帮助元数据';
        this.localeService = this.injector.get(LocaleService);
        this.lookupTitle = this.localeService.getValue('lookupConfig.lookupTitle');
        this.modalSure = this.localeService.getValue('lookupConfig.sure');
        this.modalCancel = this.localeService.getValue('lookupConfig.cancel');
        this.successSave = this.localeService.getValue('lookupConfig.successSave');
        this.helpIdError = this.localeService.getValue('lookupConfig.helpIdError');
        // this.injectService();
    }
    /**
     * @param {?} option
     * @return {?}
     */
    LookupConfigService.prototype.showConfig = /**
     * @param {?} option
     * @return {?}
     */
    function (option) {
        var _this = this;
        /** @type {?} */
        var LookupConfigF = this.cfr.resolveComponentFactory(LookupConfigComponent);
        /** @type {?} */
        var LookupConfigR = LookupConfigF.create(this.injector);
        LookupConfigR.instance.configId = option.configId;
        if (option.hasOwnProperty('metadataPath')) {
            LookupConfigR.instance.metadataPath = option.metadataPath;
        }
        if (option.hasOwnProperty('readonly')) {
            LookupConfigR.instance.readonly = option.readonly;
        }
        // if (option.hasOwnProperty('isDeisgnerEnv')) {
        //   LookupConfigR.instance.isDeisgnerEnv = option.isDeisgnerEnv;
        // }
        this.lookupDlg = this.modalService.show(LookupConfigR, {
            title: this.lookupTitle,
            width: 1000,
            height: 800,
            buttons: [
                {
                    text: this.modalSure, cls: 'btn btn-primary', handle: (/**
                     * @param {?} e
                     * @return {?}
                     */
                    function (e) {
                        _this.lookupData = LookupConfigR.instance.getdata();
                        if (_this.lookupData.helpId === '') {
                            _this.notifyService.error(_this.helpIdError);
                        }
                        else {
                            /** @type {?} */
                            var params = _this.configDataInit(_this.lookupData);
                            _this.configService.saveHelpConfig(params).subscribe((/**
                             * @param {?} metadata
                             * @return {?}
                             */
                            function (metadata) {
                                if (metadata) {
                                    _this.lookupconfigData.next(metadata['id']);
                                    _this.lookupconfigDatas.next(metadata);
                                    _this.notifyService.success(_this.successSave);
                                }
                                _this.lookupDlg.close();
                            }), (/**
                             * @param {?} error
                             * @return {?}
                             */
                            function (error) {
                                if (error.message) {
                                    _this.notifyService.error(error.message);
                                }
                            }));
                            // this.lookupconfigData.next(this.lookupData);
                        }
                    })
                },
                {
                    text: this.modalCancel, cls: 'btn btn-secondary', handle: (/**
                     * @param {?} e
                     * @return {?}
                     */
                    function (e) {
                        _this.lookupDlg.close();
                    })
                }
            ],
            showButtons: true
        });
        // return this.el;
    };
    /**
     * @return {?}
     */
    LookupConfigService.prototype.getLookupData = /**
     * @return {?}
     */
    function () {
        return this.lookupconfigData;
    };
    /**
     * @return {?}
     */
    LookupConfigService.prototype.getLookUpDatas = /**
     * @return {?}
     */
    function () {
        return this.lookupconfigDatas;
    };
    /**
     * @param {?} lookupData
     * @return {?}
     */
    LookupConfigService.prototype.configDataInit = /**
     * @param {?} lookupData
     * @return {?}
     */
    function (lookupData) {
        /** @type {?} */
        var data = {
            id: lookupData.configId,
            name: lookupData.name,
            code: lookupData.code,
            helpId: lookupData.helpId,
            queryParam: {
                filter: {
                    isUsePagination: false,
                    filterConditions: lookupData.filter
                },
                textField: lookupData.textField,
                loadTreeDataType: lookupData.loadDataType
            }
        };
        return data;
    };
    LookupConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    LookupConfigService.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: Injector },
        { type: BsModalService },
        { type: NotifyService },
        { type: ConfigService },
        { type: HttpClient }
    ]; };
    /** @nocollapse */ LookupConfigService.ngInjectableDef = i0.defineInjectable({ factory: function LookupConfigService_Factory() { return new LookupConfigService(i0.inject(i0.ComponentFactoryResolver), i0.inject(i0.INJECTOR), i0.inject(i1.BsModalService), i0.inject(i2.NotifyService), i0.inject(i3.ConfigService), i0.inject(i4.HttpClient)); }, token: LookupConfigService, providedIn: "root" });
    return LookupConfigService;
}());
export { LookupConfigService };
if (false) {
    /** @type {?} */
    LookupConfigService.prototype.lookupDlg;
    /**
     * @type {?}
     * @private
     */
    LookupConfigService.prototype.lookupconfigData;
    /**
     * @type {?}
     * @private
     */
    LookupConfigService.prototype.lookupconfigDatas;
    /** @type {?} */
    LookupConfigService.prototype.lookupData;
    /** @type {?} */
    LookupConfigService.prototype.localeService;
    /** @type {?} */
    LookupConfigService.prototype.lookupTitle;
    /** @type {?} */
    LookupConfigService.prototype.modalSure;
    /** @type {?} */
    LookupConfigService.prototype.modalCancel;
    /** @type {?} */
    LookupConfigService.prototype.successSave;
    /** @type {?} */
    LookupConfigService.prototype.helpIdError;
    /**
     * @type {?}
     * @private
     */
    LookupConfigService.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    LookupConfigService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    LookupConfigService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    LookupConfigService.prototype.notifyService;
    /**
     * @type {?}
     * @private
     */
    LookupConfigService.prototype.configService;
    /** @type {?} */
    LookupConfigService.prototype.http;
}
