// @farris/lookup-config/esm5/lib/lookup-config.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/lookup-config.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { LookupConfigComponent } from './lookup-config.component';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { FarrisButtonModule } from '@farris/ui-button';
import { ModalModule, BsModalService } from '@farris/ui-modal';
import { LookupConfigService } from './lookup-config.service';
import { ComponentLoaderFactory } from '@farris/ui-modal/component-loader';
import { PositioningService } from '@farris/ui-modal/positioning';
import { InputGroupModule } from '@farris/ui-input-group';
import { FilterModule, FilterService } from '@farris/ui-filter-editor';
import { LookupModule } from '@farris/ui-lookup';
import { LookupConfigMetaDataComponent } from './components/lookup-config-metadata.component';
import { GridModule } from '@progress/kendo-angular-grid';
import { InputsModule } from '@progress/kendo-angular-inputs';
import { NotifyModule } from '@farris/ui-notify';
import { FarrisCommonModule } from '@farris/ui-common';
import { HttpClientModule } from '@angular/common/http';
import { CommonSharedModule, HttpService, SessionService } from '@ecp-caf/caf-common';
import { ConfigService } from './config.service';
import { MetadataRTSelectModule, GSPMetadataRTServiceModule, GSPMetadataRTService } from '@gsp-lcm/metadatart-selector';
import { LoadingModule } from '@farris/ui-loading';
import { LocaleModule } from '@farris/ui-locale';
import { SelectHelpTextFieldComponent } from './components/select-help-text-field/select-help-text-field.component';
import { TreeTableModule } from '@farris/ui-treetable';
import { ComboListModule } from '@farris/ui-combo-list';
var LookupConfigModule = /** @class */ (function () {
    function LookupConfigModule() {
    }
    LookupConfigModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [LookupConfigComponent, LookupConfigMetaDataComponent, SelectHelpTextFieldComponent],
                    imports: [
                        CommonModule,
                        FormsModule,
                        ModalModule,
                        HttpClientModule,
                        FarrisDialogModule,
                        FarrisButtonModule,
                        InputGroupModule,
                        FilterModule,
                        LookupModule,
                        NotifyModule,
                        GridModule,
                        InputsModule,
                        FarrisCommonModule,
                        MetadataRTSelectModule,
                        GSPMetadataRTServiceModule.forRoot(''),
                        CommonSharedModule,
                        LoadingModule.forRoot(),
                        LocaleModule,
                        TreeTableModule,
                        ComboListModule
                    ],
                    exports: [LookupConfigComponent],
                    providers: [
                        BsModalService,
                        LookupConfigService,
                        ComponentLoaderFactory,
                        PositioningService,
                        FilterService,
                        ConfigService,
                        GSPMetadataRTService,
                        HttpService,
                        SessionService
                    ],
                    entryComponents: [LookupConfigComponent, LookupConfigMetaDataComponent, SelectHelpTextFieldComponent]
                },] }
    ];
    return LookupConfigModule;
}());
export { LookupConfigModule };
