// @gsp-wf/wf-process-editor/esm5/lib/views/time-limit-control/auto-handle-operation/auto-handle-operation.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, EventEmitter, Inject, Input, LOCALE_ID, Optional, Output, TemplateRef, ViewChild } from '@angular/core';
import { AutoHandle } from "../time-limit-control.entity";
import { MessagerService } from '@farris/ui-messager';
import { ProcessDesignerUIState } from '../../../app/process-designer.uistate';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../../i18n/pipe/wf-process-editor-locale.pipe';
var AutoHandleOperationComponent = /** @class */ (function () {
    function AutoHandleOperationComponent(message, uiState, cdr, localeId, resources, localePipe) {
        var e_1, _a;
        this.message = message;
        this.uiState = uiState;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.placeholder = '';
        this.operation = new AutoHandle('Pass', this.localePipe);
        this.submitEmitter = new EventEmitter();
        this.closeEmitter = new EventEmitter();
        this.autoHandleType = [
        // { value: 'autoPass', text: '自动通过' },
        // { value: 'autoBack', text: '自动驳回' }
        ];
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
        try {
            for (var _b = tslib_1.__values(this.uiState.propertyData.ApprovalItems), _c = _b.next(); !_c.done; _c = _b.next()) {
                var item = _c.value;
                this.autoHandleType.push({ value: item.Value, text: item.Text });
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    Object.defineProperty(AutoHandleOperationComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this.operation;
        },
        set: /**
         * @param {?} info
         * @return {?}
         */
        function (info) {
            this.operation = Object.assign({}, info);
            if (this.operation.autoHandleType === undefined) {
                /** @type {?} */
                var autoHandle = new AutoHandle(this.autoHandleType[0].value, this.localePipe);
                this.operation.autoHandleType = autoHandle.autoHandleType;
                this.operation.message = autoHandle.message;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    AutoHandleOperationComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    AutoHandleOperationComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        if (this.validator()) {
            this.submitEmitter.emit(this.operation);
        }
    };
    /**
     * @return {?}
     */
    AutoHandleOperationComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeEmitter.emit();
    };
    /**
     * @private
     * @return {?}
     */
    AutoHandleOperationComponent.prototype.validator = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.operation.message) {
            this.message.info(this.localePipe.transform('component.sendMessageOperation.opinionsNotEmpty'));
            return false;
        }
        if (!this.operation.autoHandleType) {
            this.message.info(this.localePipe.transform('component.sendMessageOperation.processingNotEmpty'));
            return false;
        }
        return true;
    };
    AutoHandleOperationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-auto-handle-operation',
                    template: "<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <div class=\"f-section-header pt-2 mx-2\">\r\n        <div class=\"f-title\">\r\n          <h4 class=\"f-title-text\">{{'component.sendMessageOperation.automaticProcessSet' | wfLocale}}</h4>\r\n        </div>\r\n      </div>\r\n      <div class=\"f-utils-fill-flex-column h-100\">\r\n        <perfect-scrollbar [config]=\"{}\">\r\n          <form class=\"h-100\">\r\n            <div class=\"farris-form form-inline farris-form-inline\">\r\n              <div class=\"farris-group-wrap\">\r\n                <div class=\"form-group farris-form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-info text-danger\">*</span>\r\n                    <span class=\"farris-label-text\">{{'component.sendMessageOperation.automaticProcessMethod' | wfLocale}}</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap ml-1\">\r\n                    <farris-combo-list [(ngModel)]=\"operation.autoHandleType\" name=\"cycleTimeUnit\" [idField]=\"'value'\" [textField]=\"'text'\"\r\n                                       [valueField]=\"'value'\" [data]=\"autoHandleType\" [panelHeight]=\"'auto'\"\r\n                                       [editable]=\"false\" [enableClear]=\"false\">\r\n                    </farris-combo-list>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n              <div class=\"farris-group-wrap my-3\">\r\n                <div class=\"form-group farris-form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-info text-danger\">*</span>\r\n                    <span class=\"farris-label-text\">{{'component.sendMessageOperation.automaticOpinionHandle' | wfLocale}}</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap ml-1\">\r\n                    <textarea class=\"form-control\" type=\"text\" value=\"message\" [(ngModel)]=\"operation.message\" name=\"message\"\r\n                              [placeholder]=\"placeholder\" style=\"height: 100px;\"></textarea>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </form>\r\n        </perfect-scrollbar>\r\n      </div>\r\n    </section>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #footer>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm mx-2\" (click)=\"clickCancel()\">{{'component.sendMessageOperation.clickCancel' | wfLocale}}</button>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">{{'component.sendMessageOperation.clickOK' | wfLocale}}</button>\r\n</ng-template>\r\n",
                    providers: [
                        WfLocalePipe,
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    AutoHandleOperationComponent.ctorParameters = function () { return [
        { type: MessagerService },
        { type: ProcessDesignerUIState },
        { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
        { type: WfLocalePipe, decorators: [{ type: Optional }] }
    ]; };
    AutoHandleOperationComponent.propDecorators = {
        footer: [{ type: ViewChild, args: ['footer',] }],
        value: [{ type: Input }],
        submitEmitter: [{ type: Output }],
        closeEmitter: [{ type: Output }]
    };
    return AutoHandleOperationComponent;
}());
export { AutoHandleOperationComponent };
if (false) {
    /** @type {?} */
    AutoHandleOperationComponent.prototype.footer;
    /** @type {?} */
    AutoHandleOperationComponent.prototype.placeholder;
    /** @type {?} */
    AutoHandleOperationComponent.prototype.operation;
    /** @type {?} */
    AutoHandleOperationComponent.prototype.submitEmitter;
    /** @type {?} */
    AutoHandleOperationComponent.prototype.closeEmitter;
    /** @type {?} */
    AutoHandleOperationComponent.prototype.autoHandleType;
    /**
     * @type {?}
     * @private
     */
    AutoHandleOperationComponent.prototype.message;
    /** @type {?} */
    AutoHandleOperationComponent.prototype.uiState;
    /**
     * @type {?}
     * @private
     */
    AutoHandleOperationComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    AutoHandleOperationComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    AutoHandleOperationComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    AutoHandleOperationComponent.prototype.localePipe;
}
