// @gsp-bef/gsp-udt-metadata/bundles/gsp-bef-gsp-udt-metadata.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@gsp-bef/gsp-cm-metadata')) :
    typeof define === 'function' && define.amd ? define('@gsp-bef/gsp-udt-metadata', ['exports', '@gsp-bef/gsp-cm-metadata'], factory) :
    (factory((global['gsp-bef'] = global['gsp-bef'] || {}, global['gsp-bef']['gsp-udt-metadata'] = {}),global.gspCmMetadata));
}(this, (function (exports,gspCmMetadata) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 统一数据类型定义
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 统一数据类型定义
     */
    var  /**
     * 统一数据类型定义
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * 统一数据类型定义
     */
    UnifiedDataTypeDef = /** @class */ (function () {
        function UnifiedDataTypeDef() {
        }
        return UnifiedDataTypeDef;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 多值业务字段
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 多值业务字段
     */
    var  /**
     * 多值业务字段
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * 多值业务字段
     */
    ComplexDataTypeDef = /** @class */ (function (_super) {
        __extends(ComplexDataTypeDef, _super);
        function ComplexDataTypeDef() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return ComplexDataTypeDef;
    }(UnifiedDataTypeDef));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 简单数据类型
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 简单数据类型
     */
    var  /**
     * 简单数据类型
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * 简单数据类型
     */
    SimpleDataTypeDef = /** @class */ (function (_super) {
        __extends(SimpleDataTypeDef, _super);
        function SimpleDataTypeDef() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            /**
             * 字段数据类型(默认string)
             */
            _this.DataType = gspCmMetadata.GSPElementDataType.String;
            /**
             * 字段对象类型(默认None)
             */
            _this.ObjectType = gspCmMetadata.GSPElementObjectType.None;
            return _this;
        }
        return SimpleDataTypeDef;
    }(UnifiedDataTypeDef));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 属性使用方式的信息
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 属性使用方式的信息[模板/约束，是否可改变使用方式]
     */
    var  /**
     * 属性使用方式的信息
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * 属性使用方式的信息[模板/约束，是否可改变使用方式]
     */
    UseTypeInfo = /** @class */ (function () {
        function UseTypeInfo() {
        }
        return UseTypeInfo;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 业务字段的json序列化结构
     */
    var /**
     * 业务字段的json序列化结构
     */ UdtMetadataContent = /** @class */ (function () {
        function UdtMetadataContent() {
        }
        return UdtMetadataContent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 关联关系
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 关联关系
     */
    var  /**
     * 关联关系
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * 关联关系
     */
    UdtAssociation = /** @class */ (function (_super) {
        __extends(UdtAssociation, _super);
        function UdtAssociation() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return UdtAssociation;
    }(gspCmMetadata.CefGSPAssociation));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    // import { Guid } from '../../../util/Guid';
    // import { GspAssociation } from '../../ref';
    // import { GSPEnumValue } from '@gsp-bef/gsp-be-metadata';
    /**
     * 字段定义
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 字段定义
     */
    var  
    // import { Guid } from '../../../util/Guid';
    // import { GspAssociation } from '../../ref';
    // import { GSPEnumValue } from '@gsp-bef/gsp-be-metadata';
    /**
     * 字段定义
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 字段定义
     */
    UdtElement = /** @class */ (function (_super) {
        __extends(UdtElement, _super);
        function UdtElement() {
            var _this = _super.call(this) || this;
            /**
             * 字段数据类型(默认string)
             */
            _this.DataType = gspCmMetadata.GSPElementDataType.String;
            // /**
            //  * 字段对象类型(默认None)
            //  */
            _this.ObjectType = gspCmMetadata.GSPElementObjectType.None;
            // this.Id = Guid.newGuid();
            _this.Length = 36;
            _this.Precision = 0;
            _this.DataType = gspCmMetadata.GSPElementDataType.String;
            _this.ObjectType = gspCmMetadata.GSPElementObjectType.None;
            _this.IsMultiLanguage = false;
            _this.IsRequire = false;
            return _this;
        }
        return UdtElement;
    }(gspCmMetadata.GspCommonField));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 属性使用方式类型
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /** @enum {string} */
    var UseType = {
        /**
         * 约束
         */
        AsConstraint: 'AsConstraint',
        /**
         * 模板
         */
        AsTemplate: 'AsTemplate',
    };
    /** @type {?} */
    var UdtTypes = [
        { text: '约束', value: UseType.AsConstraint },
        { text: '模板', value: UseType.AsTemplate }
    ];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 校验规则时机
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /** @enum {number} */
    var UdtTriggerTimePointType = {
        /// <summary>
        /// 不执行,用于判断时机比较结果
        /// </summary>
        None: 0,
        /// <summary>
        /// 数据更新后
        /// </summary>
        AfterModify: 1,
        /// <summary>
        /// 保存前
        /// </summary>
        BeforeSave: 2,
    };
    UdtTriggerTimePointType[UdtTriggerTimePointType.None] = 'None';
    UdtTriggerTimePointType[UdtTriggerTimePointType.AfterModify] = 'AfterModify';
    UdtTriggerTimePointType[UdtTriggerTimePointType.BeforeSave] = 'BeforeSave';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 字段映射方式
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /** @enum {string} */
    var ColumnMapType = {
        /**
         * 对应单列
         */
        SingleColumn: 'SingleColumn',
        /**
         * 对应多列（一一对应）
         */
        MultiColumns: 'MultiColumns',
    };
    /** @type {?} */
    var ColumnMapTypes = [
        { text: '存为多列', value: ColumnMapType.MultiColumns },
        { text: '存为一列', value: ColumnMapType.SingleColumn }
    ];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 多值业务字段数据库配置
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 多值业务字段
     */
    var  /**
     * 多值业务字段数据库配置
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * 多值业务字段
     */
    UdtDbInfo = /** @class */ (function () {
        function UdtDbInfo() {
            /**
             * 字段映射方式
             */
            this.MappingType = ColumnMapType.MultiColumns;
        }
        return UdtDbInfo;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * udt校验规则
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * udt校验规则
     */
    var  /**
     * udt校验规则
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * udt校验规则
     */
    ValidationInfo = /** @class */ (function () {
        function ValidationInfo() {
        }
        return ValidationInfo;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.ɵa = gspCmMetadata.GspCommonField;
    exports.ComplexDataTypeDef = ComplexDataTypeDef;
    exports.SimpleDataTypeDef = SimpleDataTypeDef;
    exports.UnifiedDataTypeDef = UnifiedDataTypeDef;
    exports.UseTypeInfo = UseTypeInfo;
    exports.UdtMetadataContent = UdtMetadataContent;
    exports.UdtAssociation = UdtAssociation;
    exports.UdtElement = UdtElement;
    exports.UseType = UseType;
    exports.UdtTypes = UdtTypes;
    exports.UdtTriggerTimePointType = UdtTriggerTimePointType;
    exports.ColumnMapType = ColumnMapType;
    exports.ColumnMapTypes = ColumnMapTypes;
    exports.UdtDbInfo = UdtDbInfo;
    exports.ValidationInfo = ValidationInfo;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
