// @farris/ui-tabs/esm2015/lib/tabs.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
export class TabConfig {
    constructor() {
        // 是否启用下拉框
        this.showDropDown = true;
        // 是否启用提示信息
        this.showTooltips = true;
        // 点击时滚动值
        this.scrollStep = 15;
        // 是否自动重新宽度 ---- todo:此处只处理了window的resize事件
        this.autoResize = true;
        // 下拉列表最大高度 非正值表明是不高度
        this.maxDropdownH = 200;
        // 下拉列表最大宽度： 非正值表明是不限制宽度
        this.maxDropdownW = -1;
        // 标签最大宽度：非正值表明不限制宽度
        this.maxTabW = 100;
        // 提示信息条数
        this.maxTooltipLen = 5;
    }
}
if (false) {
    /** @type {?} */
    TabConfig.prototype.showDropDown;
    /** @type {?} */
    TabConfig.prototype.showTooltips;
    /** @type {?} */
    TabConfig.prototype.scrollStep;
    /** @type {?} */
    TabConfig.prototype.autoResize;
    /** @type {?} */
    TabConfig.prototype.maxDropdownH;
    /** @type {?} */
    TabConfig.prototype.maxDropdownW;
    /** @type {?} */
    TabConfig.prototype.maxTabW;
    /** @type {?} */
    TabConfig.prototype.maxTooltipLen;
}
export class TabService {
    /**
     * @param {?} exConfig
     */
    constructor(exConfig) {
        this.config = new TabConfig();
        if (exConfig) {
            Object.assign(this.config, exConfig);
        }
    }
    /**
     * @return {?}
     */
    getConfig() {
        return this.config;
    }
}
TabService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
TabService.ctorParameters = () => [
    { type: TabConfig, decorators: [{ type: Optional }] }
];
if (false) {
    /** @type {?} */
    TabService.prototype.config;
}
