// @gsp-wf/wf-process-model/esm5/lib/model/participants/ActivityExecutorParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { ParticipantExpression } from "./ParticipantExpression";
var ActivityExecutorParticipantExpression = /** @class */ (function (_super) {
    tslib_1.__extends(ActivityExecutorParticipantExpression, _super);
    function ActivityExecutorParticipantExpression(model, activityDefId, activityDefName) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = ActivityExecutorParticipantExpression.Clr_Type_ID;
        _this.activityDefId = activityDefId;
        _this.activityDefName = activityDefName;
        return _this;
    }
    /**
     * @return {?}
     */
    ActivityExecutorParticipantExpression.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "ActivityDefId", this.activityDefId);
        BpmnModelHelper.SetAttributeValue(obj, "ActivityDefName", this.activityDefName);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ActivityExecutorParticipantExpression.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.activityDefId = ExtensionMethod.GetAttributeValue4String(obj, "ActivityDefId");
        this.activityDefName = ExtensionMethod.GetAttributeValue4String(obj, "ActivityDefName");
    };
    /**
     * @return {?}
     */
    ActivityExecutorParticipantExpression.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new ActivityExecutorParticipantExpression(this.Model);
    };
    /**
     * @return {?}
     */
    ActivityExecutorParticipantExpression.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var activityExecutorParticipant = new ActivityExecutorParticipantExpression(this.Model);
        activityExecutorParticipant.id = BpmnModelHelper.GenerateElementId();
        activityExecutorParticipant.name = this.name;
        activityExecutorParticipant.activityDefId = this.activityDefId;
        activityExecutorParticipant.activityDefName = this.activityDefName;
        // if (this.Condition) {
        //     activityExecutorParticipant.Condition = this.Condition.clone();
        // }
        return activityExecutorParticipant;
    };
    ActivityExecutorParticipantExpression.Clr_Type_ID = 'wf.ActivityExecutorParticipantExpression';
    return ActivityExecutorParticipantExpression;
}(ParticipantExpression));
export { ActivityExecutorParticipantExpression };
if (false) {
    /** @type {?} */
    ActivityExecutorParticipantExpression.Clr_Type_ID;
    /** @type {?} */
    ActivityExecutorParticipantExpression.prototype.ClrTypeID;
    /** @type {?} */
    ActivityExecutorParticipantExpression.prototype.activityDefId;
    /** @type {?} */
    ActivityExecutorParticipantExpression.prototype.activityDefName;
}
BpmnModelHelper.RegisterElementType(new ActivityExecutorParticipantExpression(new BpmnModel()));
