// @gsp-wf/wf-process-model/esm5/lib/model/participants/ComponentParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { SerializeUtils } from "../utils/SerializeUtils";
import { ParticipantExpression } from "./ParticipantExpression";
var ComponentParticipantExpression = /** @class */ (function (_super) {
    tslib_1.__extends(ComponentParticipantExpression, _super);
    function ComponentParticipantExpression(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = ComponentParticipantExpression.Clr_Type_ID;
        return _this;
    }
    /**
     * @return {?}
     */
    ComponentParticipantExpression.prototype.ResetBeforeLoadFrom = /**
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.actualParameters = new Array();
    };
    /**
     * @return {?}
     */
    ComponentParticipantExpression.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "ComponentId", this.componentId);
        BpmnModelHelper.SetAttributeValue(obj, "ComponentName", this.componentName);
        BpmnModelHelper.SetAttributeValue(obj, "MethodCode", this.methodCode);
        BpmnModelHelper.SetAttributeValue(obj, "MethodName", this.methodName);
        SerializeUtils.setListAttributeValue(obj, 'ActualParameters', this.actualParameters);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ComponentParticipantExpression.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.componentId = obj.ComponentId;
        this.componentName = obj.ComponentName;
        this.methodCode = obj.MethodCode;
        this.methodName = obj.MethodName;
        this.actualParameters = SerializeUtils.getListFromJsonObject(this.Model, obj, 'ActualParameters');
    };
    /**
     * @return {?}
     */
    ComponentParticipantExpression.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new ComponentParticipantExpression(this.Model);
    };
    /**
     * @return {?}
     */
    ComponentParticipantExpression.prototype.clone = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        /** @type {?} */
        var componentParticipant = new ComponentParticipantExpression(this.Model);
        componentParticipant.id = BpmnModelHelper.GenerateElementId();
        componentParticipant.name = this.name;
        componentParticipant.componentId = this.componentId;
        componentParticipant.componentName = this.componentName;
        componentParticipant.methodCode = this.methodCode;
        componentParticipant.methodName = this.methodName;
        componentParticipant.actualParameters = new Array();
        if (this.actualParameters && this.actualParameters.length > 0) {
            try {
                for (var _b = tslib_1.__values(this.actualParameters), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var p = _c.value;
                    /** @type {?} */
                    var parameter = p.clone();
                    componentParticipant.actualParameters.push(parameter);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        // if (this.Condition) {
        //     componentParticipant.Condition = this.Condition.clone();
        // }
        return componentParticipant;
    };
    ComponentParticipantExpression.Clr_Type_ID = 'wf.ComponentParticipantExpression';
    ComponentParticipantExpression.CLR_TYPE_NAME = '构件';
    return ComponentParticipantExpression;
}(ParticipantExpression));
export { ComponentParticipantExpression };
if (false) {
    /** @type {?} */
    ComponentParticipantExpression.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    ComponentParticipantExpression.CLR_TYPE_NAME;
    /** @type {?} */
    ComponentParticipantExpression.prototype.ClrTypeID;
    /** @type {?} */
    ComponentParticipantExpression.prototype.componentId;
    /** @type {?} */
    ComponentParticipantExpression.prototype.componentName;
    /** @type {?} */
    ComponentParticipantExpression.prototype.methodCode;
    /** @type {?} */
    ComponentParticipantExpression.prototype.methodName;
    /** @type {?} */
    ComponentParticipantExpression.prototype.actualParameters;
}
BpmnModelHelper.RegisterElementType(new ComponentParticipantExpression(new BpmnModel()));
