// @gsp-wf/wf-process-model/esm5/lib/model/participants/ParticipantScheme.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionElement } from "@edp-pmf/bpmn-model";
import { Condition } from "../extension/condition-expression/Condition";
var ParticipantScheme = /** @class */ (function (_super) {
    tslib_1.__extends(ParticipantScheme, _super);
    function ParticipantScheme(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = ParticipantScheme.Clr_Type_ID;
        _this.MexTag = 'participantScheme';
        return _this;
    }
    /**
     * @return {?}
     */
    ParticipantScheme.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new ParticipantScheme(this.Model);
    };
    /**
     * @return {?}
     */
    ParticipantScheme.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "Id", this.Id);
        BpmnModelHelper.SetAttributeValue(obj, "Name", this.Name);
        BpmnModelHelper.SetAttributeValue(obj, "Performers", this.Performers);
        if (this.Condition) {
            BpmnModelHelper.SetAttributeValue(obj, "Condition", this.Condition.ConvertToJsonObject());
        }
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ParticipantScheme.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Id = obj.Id;
        this.Name = obj.Name;
        this.Performers = obj.Performers;
        /** @type {?} */
        var conditionElement = obj.Condition;
        if (conditionElement) {
            /** @type {?} */
            var condition = new Condition(this.Model);
            condition.LoadFromJsonObject(conditionElement);
            this.Condition = condition;
        }
    };
    /**
     * @protected
     * @return {?}
     */
    ParticipantScheme.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Id = '';
        this.Name = '';
    };
    /**
     * @return {?}
     */
    ParticipantScheme.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var participantScheme = new ParticipantScheme(this.Model);
        participantScheme.Id = this.Id;
        participantScheme.Name = this.Name;
        participantScheme.Performers = this.Performers;
        if (this.Condition) {
            participantScheme.Condition = this.Condition.clone();
        }
        return participantScheme;
    };
    ParticipantScheme.Clr_Type_ID = "wf.ParticipantScheme";
    return ParticipantScheme;
}(ExtensionElement));
export { ParticipantScheme };
if (false) {
    /** @type {?} */
    ParticipantScheme.Clr_Type_ID;
    /** @type {?} */
    ParticipantScheme.prototype.ClrTypeID;
    /** @type {?} */
    ParticipantScheme.prototype.MexTag;
    /** @type {?} */
    ParticipantScheme.prototype.Id;
    /** @type {?} */
    ParticipantScheme.prototype.Name;
    /** @type {?} */
    ParticipantScheme.prototype.Performers;
    /** @type {?} */
    ParticipantScheme.prototype.Condition;
}
BpmnModelHelper.RegisterElementType(new ParticipantScheme(new BpmnModel()));
