// @gsp-wf/wf-process-model/esm5/lib/model/participants/ParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { ActionScope } from "../ActionScope";
import { Condition } from "../extension";
/**
 * @abstract
 */
var /**
 * @abstract
 */
ParticipantExpression = /** @class */ (function (_super) {
    tslib_1.__extends(ParticipantExpression, _super);
    function ParticipantExpression(model) {
        var _this = _super.call(this, model) || this;
        _this.actionScope = ActionScope.Default;
        return _this;
    }
    /**
     * @return {?}
     */
    ParticipantExpression.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "ActionScope", this.actionScope ? this.actionScope : ActionScope.Default);
        ExtensionMethod.SetAttributeValue(obj, "Id", this.id);
        ExtensionMethod.SetAttributeValue(obj, "Name", this.name);
        ExtensionMethod.SetAttributeValue(obj, "Order", this.order);
        if (this.condition) {
            ExtensionMethod.SetAttributeValue(obj, "Condition", this.condition.ConvertToJsonObject());
        }
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ParticipantExpression.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.actionScope = (/** @type {?} */ ((ExtensionMethod.GetAttributeValue4String(obj, "ActionScope") ? ExtensionMethod.GetAttributeValue4String(obj, "ActionScope") : ActionScope.Default)));
        this.id = ExtensionMethod.GetAttributeValue4String(obj, "Id");
        this.name = ExtensionMethod.GetAttributeValue4String(obj, "Name");
        this.order = ExtensionMethod.GetAttributeValue4Int64(obj, "Order");
        if (obj.Condition) {
            /** @type {?} */
            var condition = new Condition(this.Model);
            condition.LoadFromJsonObject(obj.Condition);
            this.condition = condition;
        }
    };
    return ParticipantExpression;
}(ExtensionElement));
/**
 * @abstract
 */
export { ParticipantExpression };
if (false) {
    /** @type {?} */
    ParticipantExpression.prototype.id;
    /** @type {?} */
    ParticipantExpression.prototype.name;
    /** @type {?} */
    ParticipantExpression.prototype.actionScope;
    /** @type {?} */
    ParticipantExpression.prototype.condition;
    /** @type {?} */
    ParticipantExpression.prototype.order;
}
