// @gsp-wf/wf-process-model/esm5/lib/model/participants/PositionParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { ParticipantExpression } from "./ParticipantExpression";
var PositionParticipantExpression = /** @class */ (function (_super) {
    tslib_1.__extends(PositionParticipantExpression, _super);
    function PositionParticipantExpression(model, positionId, positionName) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = PositionParticipantExpression.Clr_Type_ID;
        /// <summary>
        /// 组织类型Code
        /// </summary>
        _this.OrgType = 'Sys';
        _this.PositionId = positionId;
        _this.PositionName = positionName;
        return _this;
    }
    /**
     * @return {?}
     */
    PositionParticipantExpression.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "PositionId", this.PositionId);
        BpmnModelHelper.SetAttributeValue(obj, "PositionName", this.PositionName);
        BpmnModelHelper.SetAttributeValue(obj, "PositionType", this.PositionType);
        BpmnModelHelper.SetAttributeValue(obj, "OrgId", this.OrgId);
        BpmnModelHelper.SetAttributeValue(obj, "OrgName", this.OrgName);
        BpmnModelHelper.SetAttributeValue(obj, "OrgType", this.OrgType);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    PositionParticipantExpression.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.PositionId = obj.PositionId;
        this.PositionName = obj.PositionName;
        this.PositionType = obj.PositionType;
        this.OrgId = obj.OrgId;
        this.OrgName = obj.OrgName;
        this.OrgType = obj.OrgType;
    };
    /**
     * @return {?}
     */
    PositionParticipantExpression.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new PositionParticipantExpression(this.Model);
    };
    /**
     * @return {?}
     */
    PositionParticipantExpression.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var positionParticipant = new PositionParticipantExpression(this.Model);
        positionParticipant.id = BpmnModelHelper.GenerateElementId();
        positionParticipant.name = this.name;
        positionParticipant.PositionId = this.PositionId;
        positionParticipant.PositionName = this.PositionName;
        positionParticipant.PositionType = this.PositionType;
        positionParticipant.OrgType = this.OrgType;
        positionParticipant.OrgId = this.OrgId;
        positionParticipant.OrgName = this.OrgName;
        // if (this.Condition) {
        //     positionParticipant.Condition = this.Condition.clone();
        // }
        return positionParticipant;
    };
    PositionParticipantExpression.Clr_Type_ID = 'wf.PositionParticipantExpression';
    PositionParticipantExpression.CLR_TYPE_NAME = '岗位';
    return PositionParticipantExpression;
}(ParticipantExpression));
export { PositionParticipantExpression };
if (false) {
    /** @type {?} */
    PositionParticipantExpression.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    PositionParticipantExpression.CLR_TYPE_NAME;
    /** @type {?} */
    PositionParticipantExpression.prototype.ClrTypeID;
    /** @type {?} */
    PositionParticipantExpression.prototype.PositionId;
    /** @type {?} */
    PositionParticipantExpression.prototype.PositionName;
    /** @type {?} */
    PositionParticipantExpression.prototype.PositionType;
    /** @type {?} */
    PositionParticipantExpression.prototype.OrgId;
    /** @type {?} */
    PositionParticipantExpression.prototype.OrgName;
    /** @type {?} */
    PositionParticipantExpression.prototype.OrgType;
}
BpmnModelHelper.RegisterElementType(new PositionParticipantExpression(new BpmnModel()));
