// @gsp-wf/wf-process-model/esm5/lib/model/participants/ReportingLineParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { ParticipantExpression } from "./ParticipantExpression";
var ReportingLineParticipantExpression = /** @class */ (function (_super) {
    tslib_1.__extends(ReportingLineParticipantExpression, _super);
    function ReportingLineParticipantExpression(model, RefUserId, RefEmployeeId, RefOrgId, ReportingLineDefId, ReportingLineDefName, Dimension) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = ReportingLineParticipantExpression.Clr_Type_ID;
        _this.RefUserId = RefUserId;
        _this.RefEmployeeId = RefEmployeeId;
        _this.RefOrgId = RefOrgId;
        _this.ReportingLineDefId = ReportingLineDefId;
        _this.ReportingLineDefName = ReportingLineDefName;
        _this.Dimension = Dimension;
        return _this;
    }
    /**
     * @return {?}
     */
    ReportingLineParticipantExpression.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "RefUserId", this.RefUserId);
        BpmnModelHelper.SetAttributeValue(obj, "RefEmployeeId", this.RefEmployeeId);
        BpmnModelHelper.SetAttributeValue(obj, "Dimension", this.Dimension);
        BpmnModelHelper.SetAttributeValue(obj, "RefOrgId", this.RefOrgId);
        BpmnModelHelper.SetAttributeValue(obj, "ReportingLineDefId", this.ReportingLineDefId);
        BpmnModelHelper.SetAttributeValue(obj, "ReportingLineDefName", this.ReportingLineDefName);
        BpmnModelHelper.SetAttributeValue(obj, "Tag", this.Tag);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ReportingLineParticipantExpression.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.RefUserId = obj.RefUserId;
        this.RefEmployeeId = obj.RefEmployeeId;
        this.Dimension = obj.Dimension;
        this.RefOrgId = obj.RefOrgId;
        this.ReportingLineDefId = obj.ReportingLineDefId;
        this.ReportingLineDefName = obj.ReportingLineDefName;
        this.Tag = obj.Tag;
    };
    /**
     * @return {?}
     */
    ReportingLineParticipantExpression.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new ReportingLineParticipantExpression(this.Model);
    };
    /**
     * @return {?}
     */
    ReportingLineParticipantExpression.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var reportingLineParticipant = new ReportingLineParticipantExpression(this.Model);
        reportingLineParticipant.id = BpmnModelHelper.GenerateElementId();
        reportingLineParticipant.name = this.name;
        reportingLineParticipant.RefUserId = this.RefUserId;
        reportingLineParticipant.RefEmployeeId = this.RefEmployeeId;
        reportingLineParticipant.Dimension = this.Dimension;
        reportingLineParticipant.RefOrgId = this.RefOrgId;
        reportingLineParticipant.ReportingLineDefId = this.ReportingLineDefId;
        reportingLineParticipant.ReportingLineDefName = this.ReportingLineDefName;
        reportingLineParticipant.Tag = this.Tag;
        // if (this.Condition) {
        //     reportingLineParticipant.Condition = this.Condition.clone();
        // }
        return reportingLineParticipant;
    };
    ReportingLineParticipantExpression.Clr_Type_ID = 'wf.ReportingLineParticipantExpression';
    ReportingLineParticipantExpression.CLR_TYPE_NAME = '汇报关系';
    return ReportingLineParticipantExpression;
}(ParticipantExpression));
export { ReportingLineParticipantExpression };
if (false) {
    /** @type {?} */
    ReportingLineParticipantExpression.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    ReportingLineParticipantExpression.CLR_TYPE_NAME;
    /** @type {?} */
    ReportingLineParticipantExpression.prototype.ClrTypeID;
    /** @type {?} */
    ReportingLineParticipantExpression.prototype.Tag;
    /** @type {?} */
    ReportingLineParticipantExpression.prototype.ReportingLineDefId;
    /** @type {?} */
    ReportingLineParticipantExpression.prototype.ReportingLineDefName;
    /** @type {?} */
    ReportingLineParticipantExpression.prototype.RefOrgId;
    /** @type {?} */
    ReportingLineParticipantExpression.prototype.RefUserId;
    /** @type {?} */
    ReportingLineParticipantExpression.prototype.RefEmployeeId;
    /** @type {?} */
    ReportingLineParticipantExpression.prototype.Dimension;
}
BpmnModelHelper.RegisterElementType(new ReportingLineParticipantExpression(new BpmnModel()));
