// @gsp-wf/wf-process-model/esm5/lib/model/participants/UserParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { ParticipantExpression } from "./ParticipantExpression";
var UserParticipantExpression = /** @class */ (function (_super) {
    tslib_1.__extends(UserParticipantExpression, _super);
    function UserParticipantExpression(model, userId, name, orgType) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = UserParticipantExpression.Clr_Type_ID;
        _this.OrgType = 'Sys';
        _this.UserId = userId;
        _this.name = name;
        _this.OrgType = orgType ? orgType : 'Sys';
        return _this;
    }
    /**
     * @return {?}
     */
    UserParticipantExpression.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new UserParticipantExpression(this.Model);
    };
    /**
     * @return {?}
     */
    UserParticipantExpression.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "UserId", this.UserId);
        BpmnModelHelper.SetAttributeValue(obj, "OrgId", this.OrgId);
        BpmnModelHelper.SetAttributeValue(obj, "OrgName", this.OrgName);
        BpmnModelHelper.SetAttributeValue(obj, "OrgType", this.OrgType);
        BpmnModelHelper.SetAttributeValue(obj, "Tag", this.Tag);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    UserParticipantExpression.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.OrgId = obj.OrgId;
        this.OrgName = obj.OrgName;
        this.OrgType = obj.OrgType;
        this.UserId = obj.UserId;
        this.Tag = obj.Tag;
    };
    /**
     * @return {?}
     */
    UserParticipantExpression.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var userParticipant = new UserParticipantExpression(this.Model);
        userParticipant.id = BpmnModelHelper.GenerateElementId();
        userParticipant.name = this.name;
        userParticipant.OrgId = this.OrgId;
        userParticipant.OrgName = this.OrgName;
        userParticipant.OrgType = this.OrgType;
        userParticipant.UserId = this.UserId;
        userParticipant.Tag = this.Tag;
        // if (this.Condition) { 
        //     userParticipant.Condition = this.Condition.clone();
        // }
        return userParticipant;
    };
    UserParticipantExpression.Clr_Type_ID = "wf.UserParticipantExpression";
    UserParticipantExpression.CLR_TYPE_NAME = '用户';
    return UserParticipantExpression;
}(ParticipantExpression));
export { UserParticipantExpression };
if (false) {
    /** @type {?} */
    UserParticipantExpression.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    UserParticipantExpression.CLR_TYPE_NAME;
    /** @type {?} */
    UserParticipantExpression.prototype.ClrTypeID;
    /** @type {?} */
    UserParticipantExpression.prototype.Tag;
    /** @type {?} */
    UserParticipantExpression.prototype.UserId;
    /** @type {?} */
    UserParticipantExpression.prototype.OrgId;
    /** @type {?} */
    UserParticipantExpression.prototype.OrgName;
    /** @type {?} */
    UserParticipantExpression.prototype.OrgType;
}
BpmnModelHelper.RegisterElementType(new UserParticipantExpression(new BpmnModel()));
