// @gsp-wf/wf-process-model/esm5/lib/model/participants/VariableParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { ParticipantExpression } from "./ParticipantExpression";
var VariableParticipantExpression = /** @class */ (function (_super) {
    tslib_1.__extends(VariableParticipantExpression, _super);
    function VariableParticipantExpression(model, varValue, name, varType) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = VariableParticipantExpression.Clr_Type_ID;
        _this.VarValue = varValue;
        _this.name = name;
        _this.VarType = varType ? varType : 'Sys';
        return _this;
    }
    /**
     * @return {?}
     */
    VariableParticipantExpression.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new VariableParticipantExpression(this.Model);
    };
    /**
     * @return {?}
     */
    VariableParticipantExpression.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "Name", this.Name);
        BpmnModelHelper.SetAttributeValue(obj, "Tag", this.Tag);
        BpmnModelHelper.SetAttributeValue(obj, "VarValue", this.VarValue);
        BpmnModelHelper.SetAttributeValue(obj, "VarType", this.VarType);
        BpmnModelHelper.SetAttributeValue(obj, "Relation", this.Relation);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    VariableParticipantExpression.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Name = obj.Name;
        this.Tag = obj.Tag;
        this.VarValue = obj.VarValue;
        this.VarType = obj.VarType;
        this.Relation = obj.Relation;
    };
    /**
     * @return {?}
     */
    VariableParticipantExpression.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var variableParticipant = new VariableParticipantExpression(this.Model);
        variableParticipant.id = BpmnModelHelper.GenerateElementId();
        variableParticipant.Name = this.Name;
        variableParticipant.Tag = this.Tag;
        variableParticipant.VarType = this.VarType;
        variableParticipant.VarValue = this.VarValue;
        variableParticipant.Relation = this.Relation;
        return variableParticipant;
    };
    VariableParticipantExpression.Clr_Type_ID = "wf.VariableParticipantExpression";
    VariableParticipantExpression.CLR_TYPE_NAME = '变量参与者';
    return VariableParticipantExpression;
}(ParticipantExpression));
export { VariableParticipantExpression };
if (false) {
    /** @type {?} */
    VariableParticipantExpression.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    VariableParticipantExpression.CLR_TYPE_NAME;
    /** @type {?} */
    VariableParticipantExpression.prototype.ClrTypeID;
    /** @type {?} */
    VariableParticipantExpression.prototype.VarValue;
    /** @type {?} */
    VariableParticipantExpression.prototype.VarType;
    /** @type {?} */
    VariableParticipantExpression.prototype.Relation;
    /** @type {?} */
    VariableParticipantExpression.prototype.Tag;
    /** @type {?} */
    VariableParticipantExpression.prototype.Name;
}
BpmnModelHelper.RegisterElementType(new VariableParticipantExpression(new BpmnModel()));
