// @gsp-cmp/querysolution/esm2015/lib/service/edit.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/edit.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
/** @type {?} */
const itemIndex = (/**
 * @param {?} item
 * @param {?} data
 * @return {?}
 */
(item, data) => {
    for (let idx = 0; idx < data.length; idx++) {
        if (data[idx].id === item.id) {
            return idx;
        }
    }
    return -1;
});
const ɵ0 = itemIndex;
export class EditService {
    constructor() {
        this.createdItems = [];
        this.updatedItems = [];
    }
    /**
     * @param {?} item
     * @return {?}
     */
    update(item) {
        if (!this.isNew(item)) {
            /** @type {?} */
            const index = itemIndex(item, this.updatedItems);
            if (index !== -1) {
                this.updatedItems.splice(index, 1, item);
            }
            else {
                this.updatedItems.push(item);
            }
        }
        else {
            /** @type {?} */
            const index = this.createdItems.indexOf(item);
            this.createdItems.splice(index, 1, item);
        }
    }
    /**
     * @param {?} item
     * @return {?}
     */
    isNew(item) {
        return !item.id;
    }
    /**
     * @param {?} target
     * @param {?} source
     * @return {?}
     */
    assignValues(target, source) {
        Object.assign(target, source);
    }
}
EditService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    EditService.prototype.createdItems;
    /**
     * @type {?}
     * @private
     */
    EditService.prototype.updatedItems;
}
export { ɵ0 };
