// @gsp-cmp/querysolution/esm2015/lib/service/solution.cache.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/solution.cache.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
export class SolutionCacheService {
    /**
     * @return {?}
     */
    static clearSolutionCache() {
        SolutionCacheService.solutionListCache = {};
    }
    /**
     * @param {?} key
     * @return {?}
     */
    static removeSolutionCache(key) {
        delete SolutionCacheService.solutionListCache[key];
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    static setSolutionCache(key, value) {
        SolutionCacheService.solutionListCache[key] = value;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    static getSolutionCache(key) {
        return SolutionCacheService.solutionListCache[key];
    }
    /**
     * @return {?}
     */
    static getSolutionAllCache() {
        return SolutionCacheService.solutionListCache;
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    static setCurrentSolution(key, value) {
        SolutionCacheService.currentSolution[key] = value;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    static getCurrentSolution(key) {
        return SolutionCacheService.currentSolution[key];
    }
    /**
     * @param {?} key
     * @return {?}
     */
    static removeCurrentSolution(key) {
        delete SolutionCacheService.currentSolution[key];
    }
    /**
     * @return {?}
     */
    static clearCurrentSolution() {
        SolutionCacheService.currentSolution = {};
    }
}
SolutionCacheService.solutionListChanged = new Subject();
SolutionCacheService.solutionListCache = {};
SolutionCacheService.currentSolution = {};
if (false) {
    /** @type {?} */
    SolutionCacheService.solutionListChanged;
    /**
     * @type {?}
     * @private
     */
    SolutionCacheService.solutionListCache;
    /**
     * @type {?}
     * @private
     */
    SolutionCacheService.currentSolution;
}
