// @farris/ui-locale/fesm2015/farris-ui-locale.js
import { CommonUtils, FarrisCommonModule } from '@farris/ui-common';
import { Injectable, LOCALE_ID, Inject, InjectionToken, Optional, Pipe, defineInjectable, inject, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const COLLAPSE_DIRECTIVE_LOCALE = {
    expand: 'expand',
    fold: 'fold'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DATAGRID_LOCALE = {
    lineNumberTitle: 'NO.',
    emptyMessage: 'Empty Data',
    pagination: {
        previousLabel: 'Prev Page',
        nextLabel: 'Next Page',
        message: 'Total <b>{1}</b> items ',
        pagelist: {
            firstText: 'Display',
            lastText: 'items'
        }
    },
    filter: {
        title: 'Conditions',
        reset: 'Reset',
        clear: 'Clear',
        clearAll: 'Clear all conditions',
        setting: 'Settings',
        nofilter: '[ Empty ]',
        checkAll: 'Check All',
        and: 'And',
        or: 'Or',
        operators: {
            equal: 'equal',
            notEqual: 'not equal',
            greater: 'greater than',
            greaterOrEqual: 'greater than or equal',
            less: 'less than',
            lessOrEqual: 'less than or equal',
            contains: 'contains',
            notContains: 'not contains',
            like: 'contains',
            notLike: 'not contains',
            in: 'in',
            notIn: 'not in',
            empty: 'empty',
            notEmpty: 'not empty',
            null: 'null',
            notNull: 'not null'
        },
        more: 'More'
    },
    settings: {
        visible: 'Visible',
        sortting: 'Sortting',
        title: 'Column Settings',
        canchoose: 'Can choose',
        choosed: 'Choosed',
        asc: 'ASC',
        desc: 'DESC',
        cancelSort: 'Cancel sortting',
        ok: 'OK',
        cancel: 'Cancel',
        reset: 'Reset',
        conciseMode: 'Concise',
        advancedMode: 'Advanced',
        formatSetting: 'Column format',
        properties: 'Column properties',
        groupping: 'Groupping',
        allColumns: 'All',
        visibleColumns: 'Visible',
        hiddenColumns: 'Hidden',
        searchPlaceholder: 'Please enter a column name',
        checkall: 'Show or hide all',
        headeralign: 'Header alignment',
        dataalign: 'Data alignment',
        alignLeft: 'Left',
        alignCenter: 'Center',
        alignRight: 'Right',
        summarytype: 'Summary type',
        summarytext: 'Summary text',
        summaryNone: 'None',
        summarySum: 'Sum',
        summaryMax: 'Max',
        summaryMin: 'Min',
        summarCount: 'Count',
        summaryAverage: 'Average',
        grouppingField: 'Groupping field',
        moreGrouppingFieldWarningMessage: 'Up to 3 fields are set for grouping',
        grouppingSummary: 'Group total',
        addGrouppingFieldTip: 'Add groupping field',
        removeGrouppingFieldTip: 'Remove groupping field',
        grouppingSummaryType: 'Group total type',
        grouppingSummaryText: 'Group total text',
        restoreDefaultSettingsText: 'Are you sure you want to restore the default settings',
        simple: {
            title: 'Show Columns',
            tip: 'The selected fields can be displayed in the list. Drag to adjust the display order in the list.',
            count: 'show <span class="visible-cols-count">{0}</span> columns'
        }
    },
    selectionData: {
        clearAll: 'Clear all',
        tooltip: 'Click here show list.',
        currentLenth: `<b>{0}</b> items selected. `
    },
    groupRow: {
        tips: 'Drag columns here to group data.',
        removeColumn: 'Remove the group column.',
        clearTip: 'Clear all grouped fields.',
        clear: 'Empty'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const FILTER_EDITOR_LOCALE = {
    // 取消
    cancelButton: 'Cancel',
    // 确定
    okButton: 'OK',
    // 添加子句
    addWhere: 'Add',
    clear: 'Clear',
    // 置顶
    moveTop: 'Top',
    // 上移
    moveUp: 'Up',
    // 下移
    moveDown: 'Down',
    // 置底
    moveBottom: 'Bottom',
    // 左括号
    leftBrackets: 'Left Brackets',
    // 字段
    field: 'Field Name',
    // 操作符
    operator: 'Operator',
    // 值
    value: 'Value',
    valueType: 'Value type',
    expressType: {
        value: 'Value',
        express: 'Express',
        frontExpress: 'Front Express'
    },
    // 右括号
    rightBrackets: 'Right Brackets',
    // 关系
    relation: 'Relation',
    relationValue: {
        and: 'And',
        or: 'Or'
    },
    // 设计器
    designTab: 'Design',
    // 源代码
    jsonTab: 'JSON',
    // Sql预览
    sqlTab: 'Sql',
    title: 'Filter Designer',
    message: 'Are you sure you want to clear all current data?',
    validate: {
        bracket: 'The brackets do not match, please check',
        relation: 'The condition relationship is incomplete, please check',
        field: 'Condition field is not set, please check'
    }
};
/** @type {?} */
const ENUM_EDITOR_LOCALE = {
    // 取消
    cancelButton: 'Cancel',
    // 确定
    okButton: 'OK',
    // 添加子句
    addWhere: 'Add',
    clear: 'Clear',
    // 置顶
    moveTop: 'Top',
    // 上移
    moveUp: 'Up',
    // 下移
    moveDown: 'Down',
    remove: 'Remove',
    // 置底
    moveBottom: 'Bottom',
    title: 'Enum Data Designer',
    message: 'Are you sure you want to clear all current data?',
    value: 'Value',
    name: 'Name'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LOADING_LOCALE = {
    message: 'Loading ...'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LOOKUP_LOCALE = {
    placeholder: 'Please choose',
    favorites: 'Favorites',
    selected: 'Selected items',
    okText: 'OK',
    cancelText: 'Cancel',
    allColumns: 'All Columns',
    datalist: 'Data items',
    mustWriteSomething: 'Please write a key words.',
    mustChoosAdatarow: 'Please choose a data row！',
    tipText: 'Are these what you are looking for?',
    cascade: {
        enable: 'Enable cascade selection',
        disable: 'Disable cascade selection',
        up: 'Cascade up selection only',
        down: 'Cascade selection only'
    },
    favoriteInfo: {
        addFav: 'Collection success.',
        cancelFav: 'Unfavorite successfully. '
    },
    getAllChilds: 'Get all children',
    contextMenu: {
        expandall: 'Expand all',
        collapseall: 'Collapse all',
        expandByLayer: 'Expand by level',
        expand1: 'Expand to level 1',
        expand2: 'Expand to level 2',
        expand3: 'Expand to level 3',
        expand4: 'Expand to level 4',
        expand5: 'Expand to level 5',
        expand6: 'Expand to level 6',
        expand7: 'Expand to level 7',
        expand8: 'Expand to level 8',
        expand9: 'Expand to level 9'
    },
    quick: {
        notfind: 'Search content not found.',
        more: 'Show more'
    },
    configError: 'The help display column is not configured. Please check whether the help data source is configured correctly.',
    selectedInfo: {
        total: 'Selected items <span class="selected-counter mx-1">{0}</span>',
        clear: 'Cancel selected',
        remove: 'Delete ({0})',
        confirm: 'Are you sure you want to cancel all selected records?'
    },
    clearAllConditions: 'Clear all conditions.',
    anyFields: 'All'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const MESSAGER_LOCALE = {
    yes: 'Yes',
    no: 'No',
    ok: 'OK',
    cancel: 'Cancel',
    title: 'System Information',
    errorTitle: 'Error Information',
    prompt: {
        fontSize: {
            name: 'Font Size',
            small: 'Small',
            middle: 'Middle',
            big: 'Large',
            large: 'Extra Large',
            huge: 'Huge'
        }
    },
    exception: {
        expand: 'Expand',
        collapse: 'Collapse',
        happend: 'Happened Time',
        detail: 'Detail',
        copy: 'Copy details',
        copySuccess: 'Copy succeeded!',
        copyFailed: 'Replication failed!',
        roger: 'Got it.',
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const MULTI_SELECT_LOCALE = {
    leftTitle: 'Unselected',
    rightTitle: 'Selected',
    noDataMoveMessage: 'Please select the data to move.',
    shiftRight: 'Shift right',
    shiftLeft: 'Shift left',
    allShiftRight: 'Shift right all',
    allShiftLeft: 'Shift left all',
    top: 'Placed at the top',
    bottom: 'Placed at the bottom',
    shiftUp: 'Shift up',
    shiftDown: 'Shift down',
    emptyData: 'Empty data',
    filterPlaceholder: 'Filter...'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const NOTIFY_LOCALE = {
    title: 'System Information'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const PAGINATION_LOCALE = {
    message: 'Total <b>{1}</b> items ',
    totalinfo: {
        firstText: 'Total',
        lastText: 'items'
    },
    pagelist: {
        firstText: 'Display',
        lastText: 'items'
    },
    previous: 'Previous',
    next: 'next',
    goto: {
        prefix: 'go to',
        suffix: ''
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const QUERY_CONDITION_LOCALE = {
    configDialog: {
        unSelectedOptions: 'Unselected Options',
        selectedOptions: 'Selected Options',
        confirm: 'Confirm',
        cancel: 'Cancel',
        placeholder: 'Please input search keywords',
        moveUp: 'Move Up',
        moveAllUp: 'Move All Up',
        moveDown: 'Move Down',
        moveAllDown: 'Move All Down',
        moveRight: 'Move Right',
        moveAllRight: 'Move All Right',
        moveLeft: 'Move Left',
        moveAllLeft: 'Move All Left',
        pleaseSelect: 'Please select options',
        noOptionMove: 'No moveable options left',
        selectOptionUp: 'Please select options to move up',
        cannotMoveUp: 'Can\'t move up',
        selectOptionTop: 'Please select options to the top',
        optionIsTop: 'The option is at top',
        selectOptionDown: 'Please select options to move down',
        cannotMoveDown: 'Can\'t move down',
        selectOptionBottom: 'Please select options to the bottom',
        optionIsBottom: 'The option is at bottom'
    },
    container: {
        query: 'Filter',
        saveAs: 'Save as',
        save: 'Save',
        config: 'Configurations'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const QUERY_SOLUTION_LOCALE = {
    saveAsDialog: {
        queryPlanName: 'Query Plan Name',
        setAsDefault: 'Set as default',
        confirm: 'Confirm',
        cancel: 'Cancel',
        placeholder: 'Please input query solution name',
        pleaseInput: 'Please input query solution name',
        title: 'New Query Solution',
        maxLength: 'Query solution name exceeding nine characters'
    },
    manageDialog: {
        name: 'Name',
        property: 'Property',
        default: 'Default',
        operation: 'Action',
        confirm: 'Confirm',
        cancel: 'Cancel',
        planNameDuplicated: '{0} is duplicated',
        system: 'System Plan',
        personal: 'Personal Plan'
    },
    container: {
        default: 'Default',
        manage: 'Edit',
        arrowUp: 'Fold',
        arrowDown: 'Unfold',
        saveAs: 'Save As',
        empty: 'empty',
        save: 'Save',
        pleaseInput: 'Property \'formId\' is required',
        saveSuccess: 'Query Plan Saved',
        saveFail: 'Query Plan didn\'t save',
        planManage: 'Edit Query Plan',
        clear: 'Clear conditions',
        require: 'Please fill the {fields} before query',
        defaultName: 'Default Scheme',
        histroyName: 'Last Filter',
        sysPresetName: 'Sys Preset Scheme'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const RESPONSE_TOOLBAR_LOCALE = {
    more: 'More',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const SECTION_LOCALE = {
    expandLabel: 'expand',
    collapseLabel: 'collapse'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const SIDEBAR_LOCALE = {
    sidebar: 'Detail',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const SORT_EDITOR_LOCALE = {
    // 取消
    cancel: 'cancel',
    // 确定
    ok: 'ok',
    // 添加子句
    add: 'Add',
    clear: 'Clear',
    // 置顶
    moveTop: 'Top',
    // 上移
    moveUp: 'Up',
    // 下移
    moveDown: 'Down',
    // 置底
    moveBottom: 'Bottom',
    // 字段
    field: 'Field Name',
    // 操作符
    order: 'Order',
    asc: 'ASC',
    desc: 'DESC',
    title: 'Sort Editor'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TABS_LOCALE = {
    more: 'more',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TEXT_LOCALE = {
    yes: 'yes',
    no: 'no',
    zoom: 'Edit content in the dialog opened.',
    comments: {
        title: 'Common comments',
        manager: 'Management',
        empty: 'No data.'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TREETABLE_LOCALE = {
    emptyMessage: 'Empty Data',
    pagination: {
        previousLabel: 'Prev Page',
        nextLabel: 'Next Page',
        message: '{0} items per page, total {1} items.'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const AVATAR_LOCALE = {
    imgtitle: 'Amend',
    typeError: 'Type error',
    sizeError: 'Can not be larger than',
    uploadError: 'Upload Fail!',
    loadError: 'Load error.'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LIST_FILTER_LOCALE = {
    filter: 'Filter',
    confirm: 'OK',
    cancel: 'Cancel',
    reset: 'Reset'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const PROGRESS_STEP_LOCALE = {
    empty: 'Data is empty'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LANGUAGE_LABEL_LOCALE = {
    en: 'English',
    "zh-cn": 'Simplified Chinese',
    "zh-CHS": 'Simplified Chinese',
    "zh-CHT": 'Traditional Chiness',
    ok: 'OK',
    cancel: 'Cancel'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const VERIFY_DETAIL_LOCALE = {
    vertifyTypeAll: 'All',
    vertifyTypeError: 'Error',
    vertifyTypeEmpty: 'Empty'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const COLLAPSE_DIRECTIVE_LOCALE_ZHCHS = {
    expand: '展开',
    fold: '收起'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DATAGRID_LOCALE_ZHCHS = {
    lineNumberTitle: '序号',
    emptyMessage: '暂无数据',
    pagination: {
        previousLabel: '上一页',
        nextLabel: '下一页',
        message: '共 <b>{1}</b> 条 ',
        pagelist: {
            firstText: '显示',
            lastText: '条'
        },
        previous: '上一页',
        next: '下一页'
    },
    filter: {
        title: '过滤条件',
        reset: '重置',
        clear: '清空条件',
        clearAll: '清空所有条件',
        setting: '高级设置',
        nofilter: '[ 无 ]',
        checkAll: '全选',
        and: '并且',
        or: '或者',
        operators: {
            equal: '等于',
            notEqual: '不等于',
            greater: '大于',
            greaterOrEqual: '大于等于',
            less: '小于',
            lessOrEqual: '小于等于',
            contains: '包含',
            notContains: '不包含',
            like: '包含',
            notLike: '不包含',
            in: '属于',
            notIn: '不属于',
            empty: '为空',
            notEmpty: '不为空',
            null: 'null',
            notNull: '不为null'
        },
        more: '查看更多'
    },
    settings: {
        visible: '显示列',
        sortting: '列排序',
        title: '列配置',
        canchoose: '可选列',
        choosed: '已选列',
        asc: '升序',
        desc: '降序',
        cancelSort: '取消排序',
        ok: '确定',
        cancel: '取消',
        reset: '恢复默认',
        conciseMode: '简洁模式',
        advancedMode: '高级模式',
        formatSetting: '列格式',
        properties: '列属性',
        groupping: '分组',
        allColumns: '所有列',
        visibleColumns: '可见列',
        hiddenColumns: '隐藏列',
        searchPlaceholder: '请输入列名称',
        checkall: '全部显示/隐藏',
        headeralign: '表头对齐',
        dataalign: '数据对齐',
        alignLeft: '左对齐',
        alignCenter: '居中对齐',
        alignRight: '右对齐',
        summarytype: '汇总合计类型',
        summarytext: '汇总合计文本',
        summaryNone: '无',
        summarySum: '求和',
        summaryMax: '最大值',
        summaryMin: '最小值',
        summarCount: '计数',
        summaryAverage: '平均值',
        grouppingField: '分组字段',
        moreGrouppingFieldWarningMessage: '最多设置3个字段进行分组',
        // Up to 3 fields are set for grouping
        grouppingSummary: '分组合计',
        addGrouppingFieldTip: '添加分组字段',
        removeGrouppingFieldTip: '移除分组字段',
        grouppingSummaryType: '分组合计类型',
        grouppingSummaryText: '分组合计文本',
        restoreDefaultSettingsText: '确认要恢复默认设置吗?',
        simple: {
            title: '显示列',
            tip: '选中的字段可展示到列表中，拖拽可调整在列表中的展示顺序。',
            count: '已显示 <span class="visible-cols-count">{0}</span> 列'
        }
    },
    selectionData: {
        clearAll: '清空',
        tooltip: '点击显示已选记录列表',
        currentLenth: `已选择：<b>{0}</b> 条`
    },
    groupRow: {
        tips: '拖动列到这儿可进行数据分组',
        removeColumn: '移除分组列',
        clearTip: '清除所有分组字段',
        clear: '清空'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const FILTER_EDITOR_LOCALE_ZHCHS = {
    // 取消
    cancelButton: '取消',
    // 确定
    okButton: '确定',
    // 添加子句
    addWhere: '添加子句',
    clear: '清空',
    // 置顶
    moveTop: '置顶',
    // 上移
    moveUp: '上移',
    // 下移
    moveDown: '下移',
    // 置底
    moveBottom: '置底',
    // 左括号
    leftBrackets: '左括号',
    // 字段
    field: '字段',
    // 操作符
    operator: '操作符',
    // 值
    value: '值',
    // 值类型
    valueType: '值类型',
    expressType: {
        value: '值',
        express: '表达式',
        frontExpress: '表单表达式'
    },
    // 右括号
    rightBrackets: '右括号',
    // 关系
    relation: '关系',
    relationValue: {
        and: '并且',
        or: '或者'
    },
    designTab: '设计器',
    jsonTab: '源代码',
    sqlTab: 'Sql预览',
    title: '条件编辑器',
    message: '确认要清空当前所有数据吗？',
    validate: {
        bracket: '左右括号不匹配，请检查',
        relation: '条件关系不完整，请检查',
        field: '条件字段未设置，请检查'
    }
};
/** @type {?} */
const ENUM_EDITOR_LOCALE_ZHCHS = {
    // 取消
    cancelButton: '取消',
    // 确定
    okButton: '确定',
    // 添加子句
    addWhere: '添加',
    clear: '清空',
    // 置顶
    moveTop: '置顶',
    // 上移
    moveUp: '上移',
    // 下移
    moveDown: '下移',
    // 置底
    moveBottom: '置底',
    remove: '删除',
    title: '枚举数据编辑器',
    message: '确认要清空当前所有数据吗？',
    value: '值',
    name: '名称'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LOADING_LOCALE_ZHCHS = {
    message: '正在加载，请稍候...'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LOOKUP_LOCALE_ZHCHS = {
    placeholder: '请选择',
    favorites: '收藏夹',
    selected: '已选数据',
    okText: '确定',
    cancelText: '取消',
    allColumns: '所有列',
    datalist: '数据列表',
    mustWriteSomething: '请输入关键字后查询。',
    mustChoosAdatarow: '请选择一条记录！',
    tipText: '您要找的是不是这些？',
    cascade: {
        enable: '同步选择',
        disable: '仅选择自身',
        up: '包含上级',
        down: '包含下级'
    },
    favoriteInfo: {
        addFav: '收藏成功。',
        cancelFav: '取消收藏成功。 '
    },
    getAllChilds: '获取所有子级数据',
    contextMenu: {
        expandall: '全部展开',
        collapseall: '全部收起',
        expandByLayer: '按层级展开',
        expand1: '展开 1 级',
        expand2: '展开 2 级',
        expand3: '展开 3 级',
        expand4: '展开 4 级',
        expand5: '展开 5 级',
        expand6: '展开 6 级',
        expand7: '展开 7 级',
        expand8: '展开 8 级',
        expand9: '展开 9 级'
    },
    quick: {
        notfind: '未找到搜索内容',
        more: '显示更多'
    },
    configError: '帮助显示列未配置，请检查是否已正确配置帮助数据源! ',
    selectedInfo: {
        total: '已选 <span class="selected-counter mx-1">{0}</span> 条',
        clear: '取消已选',
        remove: '移除 ({0})',
        confirm: '您确认要取消所有选中记录吗?'
    },
    clearAllConditions: '清除所有查询条件',
    anyFields: '全部'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const MESSAGER_LOCALE_ZHCHS = {
    yes: '是',
    no: '否',
    ok: '确定',
    cancel: '取消',
    title: '系统提示',
    errorTitle: '错误提示',
    prompt: {
        fontSize: {
            name: '字体大小',
            small: '小',
            middle: '中',
            big: '大',
            large: '特大',
            huge: '超大'
        },
        tips: {
            surplus: '还可以输入 {0} 个字符',
            length: '已输入 {0} 个字符'
        }
    },
    exception: {
        expand: '展开',
        collapse: '收起',
        happend: '发生时间',
        detail: '详细信息',
        copy: '复制详细信息',
        copySuccess: '复制成功',
        copyFailed: '复制失败',
        roger: '知道了'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const MULTI_SELECT_LOCALE_ZHCHS = {
    leftTitle: '未选择',
    rightTitle: '已选择',
    noDataMoveMessage: '请选择要移动的数据。',
    shiftRight: '右移',
    shiftLeft: '左移',
    allShiftRight: '全部右移',
    allShiftLeft: '全部左移',
    top: '置顶',
    bottom: '置底',
    shiftUp: '上移',
    shiftDown: '下移',
    emptyData: '暂无数据',
    filterPlaceholder: '输入筛选项名称搜索'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const NOTIFY_LOCALE_ZHCHS = {
    title: '系统提示'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const PAGINATION_LOCALE_ZHCHS = {
    message: '共 <b>{1}</b> 条 ',
    totalinfo: {
        firstText: '共',
        lastText: '条'
    },
    pagelist: {
        firstText: '每页',
        lastText: '条'
    },
    previous: '上一页',
    next: '下一页',
    goto: {
        prefix: '跳至',
        suffix: '页'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const QUERY_CONDITION_LOCALE_ZHCHS = {
    configDialog: {
        unSelectedOptions: '未选择项',
        selectedOptions: '已选择项',
        confirm: '确定',
        cancel: '取消',
        placeholder: '请输入搜索关键字',
        moveUp: '上移',
        moveAllUp: '全部上移',
        moveDown: '下移',
        moveAllDown: '全部下移',
        moveRight: '右移',
        moveAllRight: '全部右移',
        moveLeft: '左移',
        moveAllLeft: '全部左移',
        pleaseSelect: '请选择字段',
        noOptionMove: '没有可移动字段',
        selectOptionUp: '请选择上移字段',
        cannotMoveUp: '无法上移',
        selectOptionTop: '请选择置顶字段',
        optionIsTop: '字段已置顶',
        selectOptionDown: '请选择下移字段',
        cannotMoveDown: '无法下移',
        selectOptionBottom: '请选择置底字段',
        optionIsBottom: '字段已置底'
    },
    container: {
        query: '筛选',
        saveAs: '另存为',
        save: '保存',
        config: '配置'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const QUERY_SOLUTION_LOCALE_ZHCHS = {
    saveAsDialog: {
        queryPlanName: '方案名称',
        setAsDefault: '设为默认',
        confirm: '确定',
        cancel: '取消',
        placeholder: '请输入方案名称',
        pleaseInput: '请输入方案名称',
        title: '新增方案',
        maxLength: '方案名称超出九个字'
    },
    manageDialog: {
        name: '名称',
        property: '属性',
        default: '默认',
        operation: '操作',
        confirm: '确定',
        cancel: '取消',
        planNameDuplicated: '方案名称{0}重复',
        system: '系统',
        personal: '个人'
    },
    container: {
        default: '默认',
        manage: '管理',
        arrowUp: '收起',
        arrowDown: '展开',
        saveAs: '另存为方案',
        empty: '空',
        save: '保存方案',
        pleaseInput: 'formId为必传字段，请传入',
        saveSuccess: '查询方案保存成功',
        saveFail: '查询方案保存失败',
        planManage: '方案管理',
        clear: '清空',
        require: '请填写{fields}再进行筛选',
        defaultName: '默认筛选方案',
        histroyName: '上次筛选',
        sysPresetName: '系统预置方案'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const RESPONSE_TOOLBAR_LOCALE_ZHCHS = {
    more: '更多',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const SECTION_LOCALE_ZHCHS = {
    expandLabel: '展开',
    collapseLabel: '收起'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const SIDEBAR_LOCALE_ZHCHS = {
    sidebar: '详情',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const SORT_EDITOR_LOCALE_ZHCHS = {
    // 取消
    cancel: '取消',
    // 确定
    ok: '确定',
    // 添加子句
    add: '添加',
    clear: '清空',
    // 置顶
    moveTop: '置顶',
    // 上移
    moveUp: '上移',
    // 下移
    moveDown: '下移',
    // 置底
    moveBottom: ' 置底',
    // 字段
    field: '字段',
    // 排序
    order: '排序',
    asc: '升序',
    desc: '降序',
    title: '排序设置'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TABS_LOCALE_ZHCHS = {
    more: '更多',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TEXT_LOCALE_ZHCHS = {
    yes: '是',
    no: '否',
    zoom: '在打开的对话框中编辑内容',
    comments: {
        title: '常用意见',
        manager: '意见管理',
        empty: '暂无数据'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TREETABLE_LOCALE_ZHCHS = {
    emptyMessage: '暂无数据',
    pagination: {
        previousLabel: '上一页',
        nextLabel: '下一页',
        message: '每页 {0} 条记录，共 {1} 条记录。'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const AVATAR_LOCALE_ZHCHS = {
    imgtitle: '点击修改',
    typeError: '上传图片类型不正确',
    sizeError: '上传图片不能大于',
    uploadError: '图片上传失败，请重试!',
    loadError: '加载错误'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LIST_FILTER_LOCALE_ZHCHS = {
    filter: '筛选',
    confirm: '确定',
    cancel: '取消',
    reset: '清空条件'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const PROGRESS_STEP_LOCALE_ZHCHS = {
    empty: '步骤条信息为空'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LANGUAGE_LABEL_LOCALE_ZHCHS = {
    en: '英语',
    "zh-cn": '简体中文',
    "zh-CHS": '简体中文',
    "zh-CHT": '繁体中文',
    ok: '确定',
    cancel: '取消'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const VERIFY_DETAIL_ZHCHS = {
    vertifyTypeAll: '全部',
    vertifyTypeError: '错填',
    vertifyTypeEmpty: '漏填'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LISTVIEW_LOCALE = {
    emptyMessage: 'Empty Data',
    checkall: 'Check all'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const COMBO_LOCALE = {
    placeholder: 'Please Select',
    emptyMsg: 'Empty Data',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const BATCH_EDIT_DIALOG_LOCALE = {
    title: 'Batch Edit',
    appendText: 'Append new field',
    appendTextTip: 'Append more field to edit',
    okText: 'OK',
    cancelText: 'Cancel',
    field: 'Please select field',
    fieldValue: 'Please input value',
    appendTips: 'append more columns to edit.',
    selected: 'select',
    row: 'row',
    confirmTitle: 'Info',
    neverShow: 'Do not remind again',
    confirmText: 'We will update {0} rows，Are you sure?'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const PAGE_WALKER_LOCALE = {
    next: 'Next',
    prev: 'Prev',
    skip: 'Skip',
    startNow: 'Start now'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const FOOTER_LOCALE = {
    expandText: 'More Information',
    collapseText: 'More Information'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DISCUSSION_GROUP_LOCALE = {
    submit: 'Submit',
    cancel: 'Cancel',
    colleague: 'Colleague',
    all: 'For all to see',
    related: 'Relevant personnel can see',
    confirm: 'Confirm',
    reply: 'Reply',
    emptyMessage: 'Empty Data',
    placeholder: 'Please input search keywords',
    notEmpty: 'The submitted content cannot be empty',
    selectEmployee: 'Select the employee',
    next: 'Next',
    emptySelected: 'Clear',
    emptyRight: 'Select on the left',
    allOrg: 'All',
    selected: 'Selected',
    section: 'Section ',
    people: 'Per ',
    viewMore: 'View more',
    per: ' ',
    pcs: ' ',
    emptyList: 'Empty contacts',
    advancedQuery: 'Select'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TAG_LOCALE = {
    addText: 'Add',
    placeholder: 'Please input'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const NUMERIC_LOCALE = {
    placeholder: 'Please enter the number',
    range: {
        begin: 'Please enter the begin number',
        end: 'Please enter the end number'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const FILTER_PANEL_LOCALE = {
    filter: 'Filter',
    confirm: 'OK',
    cancel: 'Cancel',
    reset: 'Reset',
    advancedFilter: 'advancedFilter',
    expand: 'Expand',
    fold: 'Fold',
    last1Month: 'LastOneMonth',
    last3Month: 'LastThreeMonth',
    last6Month: 'LastSixMonth',
    pleaseInput: 'Please input ',
    searchHistory: 'Search History',
    searchResult: 'Search Result',
    intervalFilter: 'Filter by interval',
    beginPlaceHolder: 'Minimum value',
    endPlaceHolder: 'Maximum value',
    dateBeginPlaceHolder: 'Start date',
    dateEndPlaceHolder: 'End date',
    empty: 'Empty',
    clear: 'Clear',
    today: 'Today',
    yesterday: 'Yesterday',
    checkall: 'Check all',
    searchPlaceHolder: 'Please enter a name'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const SCROLLSPY_LOCALE = {
    guide: 'Nav'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LOOKUP_CONFIG_LOCALE = {
    placeholder: 'Select the form metadata',
    code: 'Code',
    name: 'Name',
    select: 'Select the help metadata',
    filter: 'Configuration conditions',
    helpidEmpty: 'HelpID cannot be empty',
    selectTitle: 'Helps with metadata selection',
    lookupTitle: 'Help configuration',
    sure: 'Confirm',
    cancel: 'Cancel',
    successSave: 'Save successfully',
    helpIdError: 'Please select the help metadata',
    fileNamePlaceholder: 'Select the help text field',
    selectFileNameTitle: 'Text field selector',
    bindingPath: 'Binding field',
    fieldError: 'The bound field does not exist!',
    textFieldLable: 'Select the textField',
    loadTypeTitle: 'Select the loadType',
    loadTypeList: {
        all: 'All load',
        layer: 'Layer load',
        default: 'Default load'
    },
    powerTitle: 'Permission setting',
    powerObjLabel: 'Permission objects',
    powerFieldLabel: 'Permission field',
    powerOperateLabel: 'Permission operate',
    linkfieldLabel: 'Help associated field',
    powerDataTitle: 'Permission object selection',
    powerFieldTitle: 'Permission field selection',
    powerOperateTitle: 'Operation selection',
    businessLable: 'Business object ',
    powerLable: 'Permission objects',
    powerError: 'Select the permission objects',
    operateError: 'Select the Operation',
    linkfieldError: 'Select permission field'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const EN_US = {
    locale: 'EN_US',
    combo: COMBO_LOCALE,
    combolist: {},
    datagrid: DATAGRID_LOCALE,
    filterEditor: FILTER_EDITOR_LOCALE,
    enumEditor: ENUM_EDITOR_LOCALE,
    lookup: LOOKUP_LOCALE,
    loading: LOADING_LOCALE,
    modal: {},
    messager: MESSAGER_LOCALE,
    notify: NOTIFY_LOCALE,
    dialog: {},
    datatable: {},
    colorPicker: {},
    numberSpinner: NUMERIC_LOCALE,
    inputGroup: {},
    treetable: TREETABLE_LOCALE,
    multiSelect: MULTI_SELECT_LOCALE,
    sortEditor: SORT_EDITOR_LOCALE,
    tabs: TABS_LOCALE,
    timePicker: {},
    wizard: {},
    tree: {},
    tooltip: {},
    listview: LISTVIEW_LOCALE,
    text: TEXT_LOCALE,
    switch: {},
    sidebar: SIDEBAR_LOCALE,
    section: SECTION_LOCALE,
    pagination: PAGINATION_LOCALE,
    responseToolbar: RESPONSE_TOOLBAR_LOCALE,
    queryCondition: QUERY_CONDITION_LOCALE,
    querySolution: QUERY_SOLUTION_LOCALE,
    collapseDirective: COLLAPSE_DIRECTIVE_LOCALE,
    avatar: AVATAR_LOCALE,
    listFilter: LIST_FILTER_LOCALE,
    progressStep: PROGRESS_STEP_LOCALE,
    languageLabel: LANGUAGE_LABEL_LOCALE,
    verifyDetail: VERIFY_DETAIL_LOCALE,
    batchEditDialog: BATCH_EDIT_DIALOG_LOCALE,
    pageWalker: PAGE_WALKER_LOCALE,
    footer: FOOTER_LOCALE,
    discussionGroup: DISCUSSION_GROUP_LOCALE,
    tag: TAG_LOCALE,
    filterPanel: FILTER_PANEL_LOCALE,
    scrollspy: SCROLLSPY_LOCALE,
    lookupConfig: LOOKUP_CONFIG_LOCALE
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LISTVIEW_LOCALE_ZHCHS = {
    emptyMessage: '暂无数据',
    checkAll: '全选'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const BATCH_EDIT_DIALOG_LOCALE_ZHCHS = {
    title: '批量编辑',
    appendText: '添加新编辑列',
    appendTextTip: '添加更多列进行批量操作',
    okText: '确定',
    cancelText: '取消',
    field: '请选择要编辑的列：',
    fieldValue: '请输入要更改的值：',
    appendTips: '添加更多列进行批量操作',
    selected: '已选',
    row: '行',
    confirmTitle: '提示',
    neverShow: '不再提示',
    confirmText: '将修改{0}行数据，确定修改吗？'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const PAGE_WALKER_ZHCHS = {
    next: '下一步',
    prev: '上一步',
    skip: '跳过',
    startNow: '立即体验'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const FOOTER_LOCALE_ZHCHS = {
    expandText: '查看更多信息',
    collapseText: '查看更多信息'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DISCUSSION_GROUP_LOCALE_ZHCHS = {
    submit: '提交',
    cancel: '取消',
    colleague: '同事',
    all: '所有人员可见',
    related: '仅相关人员可见',
    confirm: '确定',
    reply: '回复',
    emptyMessage: '暂无数据',
    placeholder: '请输入姓名搜索',
    notEmpty: '提交内容不能为空',
    selectEmployee: '选择员工',
    next: '下级',
    emptySelected: '清空已选',
    emptyRight: '请在左侧选择人员',
    allOrg: '全部组织',
    selected: '已选',
    section: '部门',
    people: '人员',
    viewMore: '查看更多',
    per: '人',
    pcs: '个',
    emptyList: '联系人为空',
    advancedQuery: '高级查询'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TAG_LOCALE_ZHCHS = {
    addText: '添加',
    placeholder: '请输入'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const NUMERIC_LOCALE_ZHCHS = {
    placeholder: '请输入数字',
    range: {
        begin: '请输入开始数字',
        end: '请输入结束数字'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const FILTER_PANEL_LOCALE_ZHCHS = {
    filter: '筛选',
    confirm: '确定',
    cancel: '取消',
    reset: '清除筛选',
    advancedFilter: '高级筛选',
    expand: '展开',
    fold: '收起',
    last1Month: '近一月',
    last3Month: '近三月',
    last6Month: '近半年',
    pleaseInput: '请先录入',
    searchHistory: '历史搜索',
    searchResult: '查询结果',
    intervalFilter: '按区间筛选',
    beginPlaceHolder: '最低值',
    endPlaceHolder: '最高值',
    dateBeginPlaceHolder: '开始日期',
    dateEndPlaceHolder: '结束日期',
    empty: '清空全部',
    clear: '清空已选',
    today: '当天',
    yesterday: '昨天',
    checkall: '全选',
    searchPlaceHolder: '请输入名称'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const SCROLLSPY_LOCALE_ZHCHS = {
    guide: '导航'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LOOKUP_CONFIG_LOCALE_ZHCHS = {
    placeholder: '选择表单元数据',
    code: '编号',
    name: '名称',
    select: '选择帮助元数据',
    filter: '配置条件',
    helpidEmpty: 'helpId不能为空',
    selectTitle: '帮助元数据选择',
    lookupTitle: '帮助配置',
    sure: '确定',
    cancel: '取消',
    successSave: '保存成功',
    helpIdError: '请选择帮助元数据',
    fileNamePlaceholder: '选择帮助文本字段',
    selectFileNameTitle: '文本字段选择器',
    bindingPath: '绑定字段',
    fieldError: '已绑定字段不存在！',
    textFieldLable: '帮助文本字段',
    loadTypeTitle: '选择加载方式',
    loadTypeList: {
        all: '全部加载',
        layer: '分层加载',
        default: '默认'
    },
    powerTitle: '权限设置',
    powerObjLabel: '权限对象',
    powerFieldLabel: '权限字段',
    powerOperateLabel: '权限操作',
    linkfieldLabel: '帮助关联字段',
    powerDataTitle: '权限对象选择',
    powerFieldTitle: '权限字段选择',
    powerOperateTitle: '操作选择',
    businessLable: '业务对象',
    powerLable: '权限对象',
    powerError: '请选择权限对象',
    operateError: '请选择操作',
    linkfieldError: '请选择权限字段'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const COMBO_LOCALE$1 = {
    placeholder: '请选择',
    emptyMsg: '暂无数据'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const ZH_CN = {
    locale: 'ZH_CN',
    combo: COMBO_LOCALE$1,
    combolist: {},
    datagrid: DATAGRID_LOCALE_ZHCHS,
    filterEditor: FILTER_EDITOR_LOCALE_ZHCHS,
    enumEditor: ENUM_EDITOR_LOCALE_ZHCHS,
    lookup: LOOKUP_LOCALE_ZHCHS,
    loading: LOADING_LOCALE_ZHCHS,
    modal: {},
    messager: MESSAGER_LOCALE_ZHCHS,
    notify: NOTIFY_LOCALE_ZHCHS,
    dialog: {},
    datatable: {},
    colorPicker: {},
    numberSpinner: NUMERIC_LOCALE_ZHCHS,
    inputGroup: {},
    sortEditor: SORT_EDITOR_LOCALE_ZHCHS,
    treetable: TREETABLE_LOCALE_ZHCHS,
    multiSelect: MULTI_SELECT_LOCALE_ZHCHS,
    tabs: TABS_LOCALE_ZHCHS,
    timePicker: {},
    wizard: {},
    tree: {},
    tooltip: {},
    listview: LISTVIEW_LOCALE_ZHCHS,
    text: TEXT_LOCALE_ZHCHS,
    switch: {},
    sidebar: SIDEBAR_LOCALE_ZHCHS,
    section: SECTION_LOCALE_ZHCHS,
    pagination: PAGINATION_LOCALE_ZHCHS,
    responseToolbar: RESPONSE_TOOLBAR_LOCALE_ZHCHS,
    queryCondition: QUERY_CONDITION_LOCALE_ZHCHS,
    querySolution: QUERY_SOLUTION_LOCALE_ZHCHS,
    collapseDirective: COLLAPSE_DIRECTIVE_LOCALE_ZHCHS,
    avatar: AVATAR_LOCALE_ZHCHS,
    listFilter: LIST_FILTER_LOCALE_ZHCHS,
    progressStep: PROGRESS_STEP_LOCALE_ZHCHS,
    languageLabel: LANGUAGE_LABEL_LOCALE_ZHCHS,
    verifyDetail: VERIFY_DETAIL_ZHCHS,
    batchEditDialog: BATCH_EDIT_DIALOG_LOCALE_ZHCHS,
    pageWalker: PAGE_WALKER_ZHCHS,
    footer: FOOTER_LOCALE_ZHCHS,
    discussionGroup: DISCUSSION_GROUP_LOCALE_ZHCHS,
    tag: TAG_LOCALE_ZHCHS,
    filterPanel: FILTER_PANEL_LOCALE_ZHCHS,
    scrollspy: SCROLLSPY_LOCALE_ZHCHS,
    lookupConfig: LOOKUP_CONFIG_LOCALE_ZHCHS
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LOOKUP_LOCALE_ZHCHT = {
    placeholder: '請選擇',
    favorites: '收藏夾',
    selected: '已選數據',
    okText: '確定',
    cancelText: '取消',
    allColumns: '所有列',
    datalist: '數據列錶',
    mustWriteSomething: '請輸入關鍵字後查詢。',
    mustChoosAdatarow: '請選擇一條記錄！',
    tipText: '您要找的是不是這些？',
    cascade: {
        enable: '同步選擇',
        disable: '僅選擇自身',
        up: '包含上級',
        down: '包含下級'
    },
    favoriteInfo: {
        addFav: '收藏成功。',
        cancelFav: '取消收藏成功。 '
    },
    getAllChilds: '獲取所有子級數據',
    contextMenu: {
        expandall: '全部展開',
        collapseall: '全部收起',
        expandByLayer: '按層級展開',
        expand1: '展開 1 級',
        expand2: '展開 2 級',
        expand3: '展開 3 級',
        expand4: '展開 4 級',
        expand5: '展開 5 級',
        expand6: '展開 6 級',
        expand7: '展開 7 級',
        expand8: '展開 8 級',
        expand9: '展開 9 級'
    },
    quick: {
        notfind: '未找到搜索內容',
        more: '顯示更多'
    },
    configError: '幫助顯示列未配置，請檢查是否已正確配置幫助數據源!',
    selectedInfo: {
        total: '已選 <span class="selected-counter mx-1">{0}</span> 條',
        clear: '取消已選',
        remove: '移除 ({0})',
        confirm: '您確認要取消所有選中記錄嗎?'
    },
    clearAllConditions: '清除所有查詢條件',
    anyFields: '全部'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DATAGRID_LOCALE_ZHCHT = {
    lineNumberTitle: '序號',
    emptyMessage: '暫無數據',
    pagination: {
        previousLabel: '上一頁',
        nextLabel: '下一頁',
        message: '共 <b>{1}</b> 條 ',
        pagelist: {
            firstText: '顯示',
            lastText: '條'
        }
    },
    filter: {
        title: '過濾條件',
        reset: '重置',
        clear: '清空條件',
        clearAll: '清空所有條件',
        setting: '高級設置',
        nofilter: '[ 無 ]',
        checkAll: '全選',
        and: '並且',
        or: '或者',
        operators: {
            equal: '等於',
            notEqual: '不等於',
            greater: '大於',
            greaterOrEqual: '大於等於',
            less: '小於',
            lessOrEqual: '小於等於',
            contains: '包含',
            notContains: '不包含',
            like: '包含',
            notLike: '不包含',
            in: '屬於',
            notIn: '不屬於',
            empty: '為空',
            notEmpty: '不為空',
            null: 'null',
            notNull: '不為null'
        },
        more: '查看更多'
    },
    settings: {
        visible: '顯示列',
        sortting: '列排序',
        title: '列配置',
        canchoose: '可選列',
        choosed: '已選列',
        asc: '升序',
        desc: '降序',
        cancelSort: '取消排序',
        ok: '確定',
        cancel: '取消',
        reset: '恢複默認',
        conciseMode: '簡潔模式',
        advancedMode: '高級模式',
        formatSetting: '列格式',
        properties: '列屬性',
        groupping: '分組',
        allColumns: '所有列',
        visibleColumns: '可見列',
        hiddenColumns: '隱藏列',
        searchPlaceholder: '請輸入列名稱',
        checkall: '全部顯示/隱藏',
        headeralign: '表頭對齊',
        dataalign: '數據對齊',
        alignLeft: '左對齊',
        alignCenter: '居中對齊',
        alignRight: '右對齊',
        summarytype: '匯總合計類型',
        summarytext: '匯總合計文本',
        summaryNone: '無',
        summarySum: '求和',
        summaryMax: '最大值',
        summaryMin: '最小值',
        summarCount: '計數',
        summaryAverage: '平均值',
        grouppingField: '分組字段',
        moreGrouppingFieldWarningMessage: '最多設置3個字段進行分組',
        // Up to 3 fields are set for grouping
        grouppingSummary: '分組合計',
        addGrouppingFieldTip: '添加分組字段',
        removeGrouppingFieldTip: '移除分組字段',
        grouppingSummaryType: '分組合計類型',
        grouppingSummaryText: '分組合計文本',
        restoreDefaultSettingsText: '確認要恢複默認設置嗎?',
        simple: {
            title: '顯示列',
            tip: '選中的字段可展示到列表中，拖拽可調整在列表中的展示順序。',
            count: '已顯示 <span class="visible-cols-count">{0}</span> 列'
        }
    },
    selectionData: {
        clearAll: '清空',
        tooltip: '點擊顯示已選記錄列錶',
        currentLenth: `已選擇：<b>{0}</b> 條`
    },
    groupRow: {
        tips: '拖動列到這兒可進行數據分組',
        removeColumn: '移除分組列',
        clearTip: '清除所有分組字段',
        clear: '清空'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const SECTION_LOCALE_ZHCHT = {
    expandLabel: '展開',
    collapseLabel: '收起'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LOADING_LOCALE_ZHCHT = {
    message: '正在加載，請稍候...'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const FILTER_EDITOR_LOCALE_ZHCHT = {
    // 取消
    cancelButton: '取消',
    // 確定
    okButton: '確定',
    // 添加子句
    addWhere: '添加子句',
    clear: '清空',
    // 置頂
    moveTop: '置頂',
    // 上移
    moveUp: '上移',
    // 下移
    moveDown: '下移',
    // 置底
    moveBottom: '置底',
    // 左括號
    leftBrackets: '左括號',
    // 字段
    field: '字段',
    // 操作符
    operator: '操作符',
    // 值
    value: '值',
    // 值類型
    valueType: '值類型',
    expressType: {
        value: '值',
        express: '錶達式',
        frontExpress: '表單表達式'
    },
    // 右括號
    rightBrackets: '右括號',
    // 關係
    relation: '關係',
    relationValue: {
        and: '並且',
        or: '或者'
    },
    designTab: '設計器',
    jsonTab: '源代碼',
    sqlTab: 'Sql預覽',
    title: '條件編輯器',
    message: '確認要清空當前所有數據嗎？',
    validate: {
        bracket: '左右括號不匹配，請檢查',
        relation: '條件關系不完整，請檢查',
        field: '條件字段未設置，請檢查'
    }
};
/** @type {?} */
const ENUM_EDITOR_LOCALE_ZHCHT = {
    // 取消
    cancelButton: '取消',
    // 確定
    okButton: '確定',
    // 添加子句
    addWhere: '添加',
    clear: '清空',
    // 置頂
    moveTop: '置頂',
    // 上移
    moveUp: '上移',
    // 下移
    moveDown: '下移',
    // 置底
    moveBottom: '置底',
    remove: '刪除',
    title: '枚舉數據編輯器',
    message: '確認要清空當前所有數據嗎？',
    value: '值',
    name: '名稱'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const MESSAGER_LOCALE_ZHCHT = {
    yes: '是',
    no: '否',
    ok: '確定',
    cancel: '取消',
    title: '係統提示',
    errorTitle: '錯誤提示',
    prompt: {
        fontSize: {
            name: '字體大小',
            small: '小',
            middle: '中',
            big: '大',
            large: '特大',
            huge: '超大'
        }
    },
    exception: {
        expand: '展開',
        collapse: '收起',
        happend: '發生時間',
        detail: '詳細信息',
        copy: '複制詳細信息',
        copySuccess: '複制成功',
        copyFailed: '複制失敗',
        roger: '知道了'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const NOTIFY_LOCALE_ZHCHT = {
    title: '係統提示'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const PAGINATION_LOCALE_ZHCHT = {
    message: '共 <b>{1}</b> 條 ',
    totalinfo: {
        firstText: '共',
        lastText: '條'
    },
    pagelist: {
        firstText: '每頁',
        lastText: '條'
    },
    previous: '上一頁',
    next: '下一頁',
    goto: {
        prefix: '跳至',
        suffix: '頁'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const SORT_EDITOR_LOCALE_ZHCHT = {
    // 取消
    cancel: '取消',
    // 確定
    ok: '確定',
    // 添加子句
    add: '添加',
    clear: '清空',
    // 置頂
    moveTop: '置頂',
    // 上移
    moveUp: '上移',
    // 下移
    moveDown: '下移',
    // 置底
    moveBottom: ' 置底',
    // 字段
    field: '字段',
    // 排序
    order: '排序',
    asc: '升序',
    desc: '降序',
    title: '排序設置'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TEXT_LOCALE_ZHCHT = {
    yes: '是',
    no: '否',
    zoom: '在打開的對話框中編輯內容',
    comments: {
        title: '常用意見',
        manager: '意見管理',
        empty: '暫無數據'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const SIDEBAR_LOCALE_ZHCHT = {
    sidebar: '詳情',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TABS_LOCALE_ZHCHT = {
    more: '更多',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const RESPONSE_TOOLBAR_LOCALE_ZHCHT = {
    more: '更多',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const MULTI_SELECT_LOCALE_ZHCHT = {
    leftTitle: '未選擇',
    rightTitle: '已選擇',
    noDataMoveMessage: '請選擇要移動的數據。',
    shiftRight: '右移',
    shiftLeft: '左移',
    allShiftRight: '全部右移',
    allShiftLeft: '全部左移',
    top: '置頂',
    bottom: '置底',
    shiftUp: '上移',
    shiftDown: '下移',
    emptyData: '暫無數據',
    filterPlaceholder: '輸入篩選項名稱搜索'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const QUERY_CONDITION_LOCALE_ZHCHT = {
    configDialog: {
        unSelectedOptions: '未選擇項',
        selectedOptions: '已選擇項',
        confirm: '確定',
        cancel: '取消',
        placeholder: '請輸入搜索關鍵字',
        moveUp: '上移',
        moveAllUp: '全部上移',
        moveDown: '下移',
        moveAllDown: '全部下移',
        moveRight: '右移',
        moveAllRight: '全部右移',
        moveLeft: '左移',
        moveAllLeft: '全部左移',
        pleaseSelect: '請選擇字段',
        noOptionMove: '冇有可移動字段',
        selectOptionUp: '請選擇上移字段',
        cannotMoveUp: '無法上移',
        selectOptionTop: '請選擇置頂字段',
        optionIsTop: '字段已置頂',
        selectOptionDown: '請選擇下移字段',
        cannotMoveDown: '無法下移',
        selectOptionBottom: '請選擇置底字段',
        optionIsBottom: '字段已置底'
    },
    container: {
        query: '篩選',
        saveAs: '另存為',
        save: '保存',
        config: '配置'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const QUERY_SOLUTION_LOCALE_ZHCHT = {
    saveAsDialog: {
        queryPlanName: '方案名稱',
        setAsDefault: '設為默認',
        confirm: '確定',
        cancel: '取消',
        placeholder: '請輸入方案名稱',
        pleaseInput: '請輸入方案名稱',
        title: '新增方案',
        maxLength: '方案名稱超出九個字'
    },
    manageDialog: {
        name: '名稱',
        property: '屬性',
        default: '默認',
        operation: '操作',
        confirm: '確定',
        cancel: '取消',
        planNameDuplicated: '方案名稱{0}重複',
        system: '係統',
        personal: '個人'
    },
    container: {
        default: '默認',
        manage: '管理',
        arrowUp: '收起',
        arrowDown: '展開',
        saveAs: '另存為方案',
        empty: '空',
        save: '保存方案',
        pleaseInput: 'formId為必傳字段，請傳入',
        saveSuccess: '查詢方案保存成功',
        saveFail: '查詢方案保存失敗',
        planManage: '方案管理',
        clear: '清空',
        require: '請填寫{fields}再進行篩選',
        defaultName: '默認篩選方案',
        histroyName: '上次篩選',
        sysPresetName: '系统预置方案'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const COLLAPSE_DIRECTIVE_LOCALE_ZHCHT = {
    expand: '展開',
    fold: '收起'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TREETABLE_LOCALE_ZHCHT = {
    emptyMessage: '暫無數據',
    pagination: {
        previousLabel: '上一頁',
        nextLabel: '下一頁',
        message: '每頁 {0} 條記錄，共 {1} 條記錄。'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const AVATAR_LOCALE_ZHCHT = {
    imgtitle: '點擊修改',
    typeError: '上傳圖片類型不正確',
    sizeError: '上傳圖片不能大於',
    uploadError: '圖片上傳失敗，請重試!',
    loadError: '加載錯誤'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LIST_FILTER_LOCALE_ZHCHT = {
    filter: '篩選',
    confirm: '確定',
    cancel: '取消',
    reset: '清空條件'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const PROGRESS_STEP_LOCALE_ZHCHT = {
    empty: '步驟條信息為空'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LANGUAGE_LABEL_LOCALE_ZHCHT = {
    en: '英語',
    "zh-cn": '簡體中文',
    "zh-CHS": '簡體中文',
    "zh-CHT": '繁體中文',
    ok: '確定',
    cancel: '取消'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const VERIFY_DETAIL_ZHCHT = {
    vertifyTypeAll: '全部',
    vertifyTypeError: '錯填',
    vertifyTypeEmpty: '漏填'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const BATCH_EDIT_DIALOG_LOCALE_ZHCHT = {
    title: '批量編輯',
    appendText: '添加新編輯列',
    appendTextTip: '添加更多列進行批量操作',
    okText: '確定',
    cancelText: '取消',
    field: '請選擇要編輯的列：',
    fieldValue: '請輸入要更改的值：',
    appendTips: '添加更多列進行批量操作',
    selected: '已選',
    row: '行',
    confirmTitle: '提示',
    neverShow: '不再提示',
    confirmText: '將修改{0}行數據，確定修改嗎？'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const PAGE_WALKER_ZHCHT = {
    next: '下一步',
    prev: '上一步',
    skip: '跳過',
    startNow: '立即體驗'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const FOOTER_LOCALE_ZHCHT = {
    expandText: '查看更多信息',
    collapseText: '查看更多信息'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DISCUSSION_GROUP_LOCALE_ZHCHT = {
    submit: '提交',
    cancel: '取消',
    colleague: '同事',
    all: '所有人員可見',
    related: '僅相關人員可見',
    confirm: '確定',
    reply: '回複',
    emptyMessage: '暫無數據',
    placeholder: '請輸入姓名搜索',
    notEmpty: '提交內容不能為空',
    selectEmployee: '選擇員工',
    next: '下級',
    emptySelected: '清空已選',
    emptyRight: '請在左側選擇人員',
    allOrg: '全部組織',
    selected: '已選',
    section: '部門',
    people: '人員',
    viewMore: '查看更多',
    per: '人',
    pcs: '個',
    emptyList: '聯係人為空',
    advancedQuery: '高級查詢'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const TAG_LOCALE_ZHCHT = {
    addText: '添加',
    placeholder: '請輸入'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const NUMERIC_LOCALE_ZHCHT = {
    placeholder: '請輸入數字',
    range: {
        begin: '請輸入開始數字',
        end: '請輸入結束數字'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const FILTER_PANEL_LOCALE_ZHCHT = {
    filter: '篩選',
    confirm: '確定',
    cancel: '取消',
    reset: '清除篩選',
    advancedFilter: '高級篩選',
    expand: '展開',
    fold: '收起',
    last1Month: '近一月',
    last3Month: '近三月',
    last6Month: '近半年',
    pleaseInput: '請先錄入',
    searchHistory: '歷史搜索',
    searchResult: '查詢結果',
    intervalFilter: '按區間篩選',
    beginPlaceHolder: '最低值',
    endPlaceHolder: '最高值',
    dateBeginPlaceHolder: '開始日期',
    dateEndPlaceHolder: '結束日期',
    empty: '清空全部',
    clear: '清空已選',
    today: '當天',
    yesterday: '昨天',
    checkall: '全選',
    searchPlaceHolder: '請輸入名稱'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const SCROLLSPY_LOCALE_ZHCHT = {
    guide: '導航'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LOOKUP_CONFIG_LOCALE_ZHCHT = {
    placeholder: '選擇表單元數據',
    code: '編號',
    name: '名稱',
    select: '選擇幫助元數據',
    filter: '配置條件',
    helpidEmpty: 'helpId不能為空',
    selectTitle: '幫助元數據選擇',
    lookupTitle: '幫助配置',
    sure: '確定',
    cancel: '取消',
    successSave: '保存成功',
    helpIdError: '請選擇幫助元數據',
    fileNamePlaceholder: '選擇幫助文本字段',
    selectFileNameTitle: '文本字段選擇器',
    bindingPath: '綁定字段',
    fieldError: '已綁定字段不存在！',
    textFieldLable: '幫助文本字段',
    loadTypeTitle: '選擇加載方式',
    loadTypeList: {
        all: '全部加載',
        layer: '分層加載',
        default: '默認'
    },
    powerTitle: '權限設置',
    powerObjLabel: '權限對象',
    powerFieldLabel: '權限字段',
    powerOperateLabel: '權限操作',
    linkfieldLabel: '幫助關聯字段',
    powerDataTitle: '權限對象選擇',
    powerFieldTitle: '權限字段選擇',
    powerOperateTitle: '操作選擇',
    businessLable: '業務對象',
    powerLable: '權限對象',
    powerError: '請選擇權限對象',
    operateError: '請選擇操作',
    linkfieldError: '請選擇權限字段'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const COMBO_LOCALE$2 = {
    placeholder: '請選擇',
    emptyMsg: '暫無數據'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const LISTVIEW_LOCALE_ZHCHT = {
    emptyMessage: '暫無數據',
    checkall: '全選'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const ZH_CHT = {
    locale: 'ZH_CHT',
    combo: COMBO_LOCALE$2,
    combolist: {},
    datagrid: DATAGRID_LOCALE_ZHCHT,
    filterEditor: FILTER_EDITOR_LOCALE_ZHCHT,
    enumEditor: ENUM_EDITOR_LOCALE_ZHCHT,
    lookup: LOOKUP_LOCALE_ZHCHT,
    loading: LOADING_LOCALE_ZHCHT,
    modal: {},
    messager: MESSAGER_LOCALE_ZHCHT,
    notify: NOTIFY_LOCALE_ZHCHT,
    dialog: {},
    datatable: {},
    colorPicker: {},
    numberSpinner: NUMERIC_LOCALE_ZHCHT,
    inputGroup: {},
    sortEditor: SORT_EDITOR_LOCALE_ZHCHT,
    treetable: TREETABLE_LOCALE_ZHCHT,
    multiSelect: MULTI_SELECT_LOCALE_ZHCHT,
    tabs: TABS_LOCALE_ZHCHT,
    timePicker: {},
    wizard: {},
    tree: {},
    tooltip: {},
    listview: LISTVIEW_LOCALE_ZHCHT,
    text: TEXT_LOCALE_ZHCHT,
    switch: {},
    sidebar: SIDEBAR_LOCALE_ZHCHT,
    section: SECTION_LOCALE_ZHCHT,
    pagination: PAGINATION_LOCALE_ZHCHT,
    responseToolbar: RESPONSE_TOOLBAR_LOCALE_ZHCHT,
    queryCondition: QUERY_CONDITION_LOCALE_ZHCHT,
    querySolution: QUERY_SOLUTION_LOCALE_ZHCHT,
    collapseDirective: COLLAPSE_DIRECTIVE_LOCALE_ZHCHT,
    avatar: AVATAR_LOCALE_ZHCHT,
    listFilter: LIST_FILTER_LOCALE_ZHCHT,
    progressStep: PROGRESS_STEP_LOCALE_ZHCHT,
    languageLabel: LANGUAGE_LABEL_LOCALE_ZHCHT,
    verifyDetail: VERIFY_DETAIL_ZHCHT,
    batchEditDialog: BATCH_EDIT_DIALOG_LOCALE_ZHCHT,
    pageWalker: PAGE_WALKER_ZHCHT,
    footer: FOOTER_LOCALE_ZHCHT,
    discussionGroup: DISCUSSION_GROUP_LOCALE_ZHCHT,
    tag: TAG_LOCALE_ZHCHT,
    filterPanel: FILTER_PANEL_LOCALE_ZHCHT,
    scrollspy: SCROLLSPY_LOCALE_ZHCHT,
    lookupConfig: LOOKUP_CONFIG_LOCALE_ZHCHT
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const FARRIS_LOCALES = {
    'zh-CHS': ZH_CN,
    en: EN_US,
    'zh-CHT': ZH_CHT
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const FARRIS_LOCAL_CUSTOM_DATA = new InjectionToken('自定义语言数据， 格式：{ "languageCode": { "name": { "key": "value" } } }');
class LocaleService {
    /**
     * @param {?} localeId
     * @param {?} utils
     * @param {?} localeData
     */
    constructor(localeId, utils, localeData) {
        this.localeId = localeId;
        this.utils = utils;
        this.localeData = localeData;
        if (!localeId) {
            localeId = 'zh-CHS';
        }
        // 合并语言资源
        this.setLocaleData(localeData);
    }
    /**
     * 获取语言资源
     *  path : 资源路径，如:
     *  ```
     *  getResources('zh-CHS') // 返回所有中文语言资源
     *  getResources('zh-CHS.lookup') // 返回所有中文下帮助的资源
     *  ```
     * @param {?=} path
     * @return {?}
     */
    getResources(path = '') {
        if (path) {
            return this.utils.getValue(path, FARRIS_LOCALES);
        }
        return FARRIS_LOCALES;
    }
    /**
     * @param {?} ctrlName
     * @return {?}
     */
    getComponentOptions(ctrlName) {
        return this.langData[ctrlName];
    }
    /**
     * @param {?} propertyName
     * @return {?}
     */
    getValue(propertyName) {
        /** @type {?} */
        const val = this.utils.getValue(propertyName, this.langData);
        if (val) {
            return val;
        }
        else {
            // console.warn(`Can\'t find language Data for the ${propertyName}`);
            return '';
        }
    }
    /**
     * 合并现有的多语资源
     * @param {?=} localeData
     * @return {?}
     */
    setLocaleData(localeData = null) {
        if (localeData) {
            // 解析得到提供的语言编码
            /** @type {?} */
            const langCodes = Object.keys(localeData);
            langCodes.forEach((/**
             * @param {?} code
             * @return {?}
             */
            (code) => {
                /** @type {?} */
                const resName = Object.keys(localeData[code]);
                resName.forEach((/**
                 * @param {?} k
                 * @return {?}
                 */
                k => {
                    this.appendLanguageResource(k, localeData[code][k], code);
                }));
            }));
        }
        this.langData = FARRIS_LOCALES[this.localeId];
        if (!this.langData) {
            this.langData = FARRIS_LOCALES['zh-CHS'];
        }
    }
    /**
     * @private
     * @param {?} key
     * @param {?} data
     * @param {?=} lang
     * @return {?}
     */
    appendLanguageResource(key, data, lang = this.localeId) {
        if (!lang) {
            this.langData[key] = this.langData[key] || {};
            this.langData[key] = Object.assign(this.langData[key], data || {});
        }
        else {
            FARRIS_LOCALES[lang][key] = data || {};
        }
    }
}
LocaleService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
LocaleService.ctorParameters = () => [
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
    { type: CommonUtils },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [FARRIS_LOCAL_CUSTOM_DATA,] }] }
];
/** @nocollapse */ LocaleService.ngInjectableDef = defineInjectable({ factory: function LocaleService_Factory() { return new LocaleService(inject(LOCALE_ID), inject(CommonUtils), inject(FARRIS_LOCAL_CUSTOM_DATA, 8)); }, token: LocaleService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class FarrisLocalePipe {
    /**
     * @param {?} localeService
     */
    constructor(localeService) {
        this.localeService = localeService;
    }
    /**
     * @param {?} name
     * @param {?=} defaultVal
     * @param {?=} customText
     * @return {?}
     */
    transform(name, defaultVal = '', customText = '') {
        if (!customText) {
            /** @type {?} */
            const r = this.localeService.getValue(name);
            return r ? r : defaultVal;
        }
        return customText;
    }
}
FarrisLocalePipe.decorators = [
    { type: Pipe, args: [{ name: 'locale' },] }
];
/** @nocollapse */
FarrisLocalePipe.ctorParameters = () => [
    { type: LocaleService }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class LocaleModule {
    /**
     * @param {?=} customData
     * @return {?}
     */
    static forRoot(customData) {
        return {
            ngModule: LocaleModule,
            providers: [
                { provide: FARRIS_LOCAL_CUSTOM_DATA, useValue: customData || {} },
                LocaleService
            ]
        };
    }
}
LocaleModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    FarrisLocalePipe
                ],
                imports: [
                    FarrisCommonModule.forRoot()
                ],
                exports: [FarrisLocalePipe],
                providers: [
                    LocaleService
                ]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { EN_US, ZH_CN, COLLAPSE_DIRECTIVE_LOCALE, DATAGRID_LOCALE, FILTER_EDITOR_LOCALE, ENUM_EDITOR_LOCALE, LOADING_LOCALE, LOOKUP_LOCALE, MESSAGER_LOCALE, MULTI_SELECT_LOCALE, NOTIFY_LOCALE, PAGINATION_LOCALE, QUERY_CONDITION_LOCALE, QUERY_SOLUTION_LOCALE, RESPONSE_TOOLBAR_LOCALE, SECTION_LOCALE, SIDEBAR_LOCALE, SORT_EDITOR_LOCALE, TABS_LOCALE, TEXT_LOCALE, TREETABLE_LOCALE, AVATAR_LOCALE, LIST_FILTER_LOCALE, PROGRESS_STEP_LOCALE, LANGUAGE_LABEL_LOCALE, VERIFY_DETAIL_LOCALE, COLLAPSE_DIRECTIVE_LOCALE_ZHCHS, DATAGRID_LOCALE_ZHCHS, FILTER_EDITOR_LOCALE_ZHCHS, ENUM_EDITOR_LOCALE_ZHCHS, LOADING_LOCALE_ZHCHS, LOOKUP_LOCALE_ZHCHS, MESSAGER_LOCALE_ZHCHS, MULTI_SELECT_LOCALE_ZHCHS, NOTIFY_LOCALE_ZHCHS, PAGINATION_LOCALE_ZHCHS, QUERY_CONDITION_LOCALE_ZHCHS, QUERY_SOLUTION_LOCALE_ZHCHS, RESPONSE_TOOLBAR_LOCALE_ZHCHS, SECTION_LOCALE_ZHCHS, SIDEBAR_LOCALE_ZHCHS, SORT_EDITOR_LOCALE_ZHCHS, TABS_LOCALE_ZHCHS, TEXT_LOCALE_ZHCHS, TREETABLE_LOCALE_ZHCHS, AVATAR_LOCALE_ZHCHS, LIST_FILTER_LOCALE_ZHCHS, PROGRESS_STEP_LOCALE_ZHCHS, LANGUAGE_LABEL_LOCALE_ZHCHS, VERIFY_DETAIL_ZHCHS, FARRIS_LOCAL_CUSTOM_DATA, LocaleService, FarrisLocalePipe, LocaleModule, FARRIS_LOCALES, BATCH_EDIT_DIALOG_LOCALE as ɵd, COMBO_LOCALE as ɵa, DISCUSSION_GROUP_LOCALE as ɵg, FILTER_PANEL_LOCALE as ɵi, FOOTER_LOCALE as ɵf, LISTVIEW_LOCALE as ɵc, LOOKUP_CONFIG_LOCALE as ɵk, NUMERIC_LOCALE as ɵb, PAGE_WALKER_LOCALE as ɵe, SCROLLSPY_LOCALE as ɵj, TAG_LOCALE as ɵh, AVATAR_LOCALE_ZHCHT as ɵbt, BATCH_EDIT_DIALOG_LOCALE_ZHCHT as ɵby, COLLAPSE_DIRECTIVE_LOCALE_ZHCHT as ɵbs, COMBO_LOCALE$2 as ɵx, DATAGRID_LOCALE_ZHCHT as ɵy, DISCUSSION_GROUP_LOCALE_ZHCHT as ɵcb, ENUM_EDITOR_LOCALE_ZHCHT as ɵba, FILTER_EDITOR_LOCALE_ZHCHT as ɵz, FILTER_PANEL_LOCALE_ZHCHT as ɵcd, FOOTER_LOCALE_ZHCHT as ɵca, ZH_CHT as ɵw, LANGUAGE_LABEL_LOCALE_ZHCHT as ɵbw, LIST_FILTER_LOCALE_ZHCHT as ɵbu, LISTVIEW_LOCALE_ZHCHT as ɵbk, LOADING_LOCALE_ZHCHT as ɵbc, LOOKUP_LOCALE_ZHCHT as ɵbb, LOOKUP_CONFIG_LOCALE_ZHCHT as ɵcf, MESSAGER_LOCALE_ZHCHT as ɵbd, MULTI_SELECT_LOCALE_ZHCHT as ɵbi, NOTIFY_LOCALE_ZHCHT as ɵbe, NUMERIC_LOCALE_ZHCHT as ɵbf, PAGE_WALKER_ZHCHT as ɵbz, PAGINATION_LOCALE_ZHCHT as ɵbo, PROGRESS_STEP_LOCALE_ZHCHT as ɵbv, QUERY_CONDITION_LOCALE_ZHCHT as ɵbq, QUERY_SOLUTION_LOCALE_ZHCHT as ɵbr, RESPONSE_TOOLBAR_LOCALE_ZHCHT as ɵbp, SCROLLSPY_LOCALE_ZHCHT as ɵce, SECTION_LOCALE_ZHCHT as ɵbn, SIDEBAR_LOCALE_ZHCHT as ɵbm, SORT_EDITOR_LOCALE_ZHCHT as ɵbg, TABS_LOCALE_ZHCHT as ɵbj, TAG_LOCALE_ZHCHT as ɵcc, TEXT_LOCALE_ZHCHT as ɵbl, TREETABLE_LOCALE_ZHCHT as ɵbh, VERIFY_DETAIL_ZHCHT as ɵbx, BATCH_EDIT_DIALOG_LOCALE_ZHCHS as ɵo, COMBO_LOCALE$1 as ɵl, DISCUSSION_GROUP_LOCALE_ZHCHS as ɵr, FILTER_PANEL_LOCALE_ZHCHS as ɵt, FOOTER_LOCALE_ZHCHS as ɵq, LISTVIEW_LOCALE_ZHCHS as ɵn, LOOKUP_CONFIG_LOCALE_ZHCHS as ɵv, NUMERIC_LOCALE_ZHCHS as ɵm, PAGE_WALKER_ZHCHS as ɵp, SCROLLSPY_LOCALE_ZHCHS as ɵu, TAG_LOCALE_ZHCHS as ɵs };
