// @gsp-cmp/ubcc-param-data-ui/esm2015/lib/ubcc-param-data-ui.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ubcc-param-data-ui.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ComponentFactoryResolver, Injector } from '@angular/core';
import { UbccParamDataUtils } from './utils/param-data-utils';
import { BsModalService } from '@farris/ui-modal';
import { UbccParamDataUiDialogComponent } from './ubcc-param-data-ui-dialog/ubcc-param-data-ui-dialog.component';
export class UbccParamDataUiService {
    /**
     * @param {?} resolver
     * @param {?} injector
     * @param {?} util
     * @param {?} modalService
     */
    constructor(resolver, injector, util, modalService) {
        this.resolver = resolver;
        this.injector = injector;
        this.util = util;
        this.modalService = modalService;
    }
    /**
     * @param {?} templateId
     * @param {?} paramData
     * @return {?}
     */
    show(templateId, paramData) {
        return this.create(templateId, paramData);
    }
    /**
     * @private
     * @param {?} templateId
     * @param {?} data
     * @return {?}
     */
    create(templateId, data) {
        /** @type {?} */
        let fac = this.resolver.resolveComponentFactory(UbccParamDataUiDialogComponent);
        /** @type {?} */
        let cmpRef = fac.create(this.injector);
        /** @type {?} */
        let modalConfig = this.util.getModalConfig(cmpRef.instance.btnRef);
        /** @type {?} */
        let dialog = this.modalService.show(cmpRef, modalConfig);
        cmpRef.instance.dialog = dialog;
        cmpRef.instance.templateId = templateId;
        cmpRef.instance.data = data;
        return { confirmEvent: cmpRef.instance.confirmEmitter, closeEvent: cmpRef.instance.closeEmitter, close: cmpRef.instance.close };
    }
}
UbccParamDataUiService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
UbccParamDataUiService.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: Injector },
    { type: UbccParamDataUtils },
    { type: BsModalService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    UbccParamDataUiService.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    UbccParamDataUiService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    UbccParamDataUiService.prototype.util;
    /**
     * @type {?}
     * @private
     */
    UbccParamDataUiService.prototype.modalService;
}
