// @farris/ui-datagrid-dff-editors/esm5/lib/datagrid-dff-editors.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/datagrid-dff-editors.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { DatagridModule } from '@farris/ui-datagrid';
import { DatagridEditorsModule } from '@farris/ui-datagrid-editors';
import { FarrisFormsModule } from '@farris/ui-forms';
import { EmployeeSelectorModule } from '@farris/ui-employee-selector';
import { AdminOrganizationSelectorModule } from '@farris/ui-adminorganization-selector';
import { DatagridDffEditorsComponent } from './datagrid-dff-editors.component';
import { DatagridEmployeeSelectorComponent } from './grid-employee-selector/grid-employee-selector.component';
import { DataGridAdminOrganizationSelectorComponent } from './grid-adminorganization-selector/grid-adminorganization-selector.component';
var DatagridDffEditorsModule = /** @class */ (function () {
    function DatagridDffEditorsModule() {
    }
    DatagridDffEditorsModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DatagridDffEditorsComponent,
                        DatagridEmployeeSelectorComponent,
                        DataGridAdminOrganizationSelectorComponent
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        DatagridModule,
                        DatagridEditorsModule,
                        FarrisFormsModule,
                        EmployeeSelectorModule,
                        AdminOrganizationSelectorModule
                    ],
                    providers: [],
                    entryComponents: [
                        DatagridEmployeeSelectorComponent,
                        DataGridAdminOrganizationSelectorComponent
                    ],
                    exports: [
                        DatagridDffEditorsComponent,
                        DatagridEmployeeSelectorComponent,
                        DataGridAdminOrganizationSelectorComponent
                    ]
                },] }
    ];
    return DatagridDffEditorsModule;
}());
export { DatagridDffEditorsModule };
