// @farris/ui-enum-editor/fesm2015/farris-ui-enum-editor.js
import { BsModalService } from '@farris/ui-modal';
import { Subject } from 'rxjs';
import { cloneDeep } from 'lodash-es';
import { InputGroupModule } from '@farris/ui-input-group';
import { NG_VALUE_ACCESSOR, FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { MessagerService, MessagerModule } from '@farris/ui-messager';
import { LocaleService, LocaleModule } from '@farris/ui-locale';
import { Component, ElementRef, Input, Output, Renderer2, ViewChild, EventEmitter, HostBinding, Injectable, Injector, ComponentFactoryResolver, NgModule, defineInjectable, inject, INJECTOR, forwardRef, ChangeDetectorRef } from '@angular/core';
import { FarrisCommonModule } from '@farris/ui-common';
import { PerfectScrollbarModule } from '@farris/ui-perfect-scrollbar';
import { LookupComponent, LookupModule } from '@farris/ui-lookup';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class EnumEditorComponent {
    /**
     * @param {?} el
     * @param {?} render
     * @param {?} messagerService
     * @param {?} localeSer
     */
    constructor(el, render, messagerService, localeSer) {
        this.el = el;
        this.render = render;
        this.messagerService = messagerService;
        this.localeSer = localeSer;
        /**
         * enum data
         */
        this.data = [];
        this.columns = [];
        this.textField = 'name';
        this.valueField = 'value';
        this.dataChange = new EventEmitter();
        this.cls = 'd-flex flex-column flex-fill';
        this.overflow = 'hidden';
        this.height = '100%';
        this.showSortBtns = true;
        this.currentEnumData = null;
        this.newEnumData = {
            [this.textField]: '',
            [this.valueField]: ''
        };
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} index
     * @return {?}
     */
    insertCondition(index) {
        /** @type {?} */
        const newEnumData = Object.assign({}, this.newEnumData);
        if (index === 0) {
            this.data.unshift(newEnumData);
        }
        else {
            this.data.splice(index, 0, newEnumData);
        }
        this.changeConditionList();
    }
    /**
     * @param {?} index
     * @return {?}
     */
    remove(index) {
        this.data.splice(index, 1);
        this.changeConditionList();
        if (this.currentEnumData) {
            if (index === this.currentEnumData.index) {
                this.currentEnumData = null;
            }
        }
    }
    /**
     * @return {?}
     */
    onAddFilter() {
        this.data = [...this.data, Object.assign({}, this.newEnumData)];
        this.changeConditionList();
        this.itemIntoView();
    }
    /**
     * @return {?}
     */
    onClear() {
        this.messagerService.question(this.localeSer.getValue('enumEditor.message'), (/**
         * @return {?}
         */
        () => {
            this.data = [];
            this.currentEnumData = null;
            this.changeConditionList();
        }));
    }
    /**
     * @return {?}
     */
    onMoveTop() {
        if (this.currentEnumData) {
            this.data.unshift(this.currentEnumData.data);
            this.data.splice(this.currentEnumData.index + 1, 1);
            this.currentEnumData.index = 0;
            this.changeConditionList();
        }
    }
    /**
     * @return {?}
     */
    onMovePrev() {
        if (this.currentEnumData) {
            /** @type {?} */
            const index = this.currentEnumData.index;
            /** @type {?} */
            const tempArr = this.data.splice(index, 1);
            this.data.splice(index - 1, 0, ...tempArr);
            this.currentEnumData.index = index - 1;
            this.changeConditionList();
        }
    }
    /**
     * @return {?}
     */
    onMoveNext() {
        if (this.currentEnumData) {
            /** @type {?} */
            const index = this.currentEnumData.index;
            /** @type {?} */
            const tempArr = this.data.splice(index, 1);
            this.data.splice(index + 1, 0, ...tempArr);
            this.currentEnumData.index = index + 1;
            this.changeConditionList();
        }
    }
    /**
     * @return {?}
     */
    onMoveBottom() {
        if (this.currentEnumData) {
            this.data.push(this.currentEnumData.data);
            this.data.splice(this.currentEnumData.index, 1);
            this.currentEnumData.index = this.data.length - 1;
            this.changeConditionList();
        }
    }
    /**
     * @private
     * @return {?}
     */
    itemIntoView() {
        if (this.tablebodybox) {
            setTimeout((/**
             * @return {?}
             */
            () => {
                /** @type {?} */
                const tr = this.tablebodybox.nativeElement.querySelector('tbody tr:last-child');
                if (tr) {
                    if (tr.scrollIntoViewIfNeeded) {
                        tr.scrollIntoViewIfNeeded(false);
                    }
                    else {
                        tr.scrollIntoView({ block: 'nearest' });
                    }
                }
            }), 50);
        }
    }
    /**
     * @private
     * @return {?}
     */
    changeConditionList() {
        this.dataChange.next(this.data);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    selected($event) {
        // console.log($event);
        this.currentEnumData = $event;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    isSelected($event) {
        if (this.currentEnumData) {
            return this.currentEnumData.index === $event;
        }
        return false;
    }
}
EnumEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'farris-enum-editor',
                template: "<!-- <div class=\"table-header \">\r\n    <table class=\"table table-bordered\" style=\"margin-bottom: 0\">\r\n        <thead>\r\n            <tr>\r\n                <th style=\"width: 42px\"></th>\r\n                <th><b>{{'enumEditor.value'| locale}}</b></th>\r\n                <th><b>{{'enumEditor.name'| locale}}</b></th>\r\n            </tr>\r\n        </thead>\r\n    </table>\r\n</div> -->\r\n<div style=\"\r\n    border-top: 1px solid  white;\r\n    position: absolute;\r\n    top: 0;\r\n    width: 100%;\r\n    z-index: 2;\r\n\"></div>\r\n<div class=\"table-body flex-fill win11Scroll show nobtn \" style=\"margin-left:14px; margin-right: 0; overflow-y: scroll;\" #tablebodybox>\r\n    <!-- <perfect-scrollbar [config]=\"{}\"> -->\r\n        <table class=\"table table-bordered\" #tbl>\r\n            <thead style=\"position: sticky; top: 0; z-index: 1;\">\r\n                <tr>\r\n                    <th style=\"width: 42px\"></th>\r\n                    <th><b>{{'enumEditor.value'| locale}}</b></th>\r\n                    <th><b>{{'enumEditor.name'| locale}}</b></th>\r\n                </tr>\r\n            </thead>\r\n            <tbody>\r\n                <tr *ngFor=\"let item of data; let index=index;\" (click)=\"selected({'index':index, 'data': item})\" [class.selected]=\"isSelected(index)\" >\r\n                    <td style=\"width: 42px\">\r\n                        <button type=\"button\" class=\"btn btn-link\" (click)=\"remove(index)\">\r\n                            <span class=\"f-icon f-icon-close\" style=\"color: red\"></span>\r\n                        </button>\r\n                    </td>\r\n                    <td>\r\n                        <input type=\"text\" style=\"width:100%\" autocomplete=\"off\" placeholder=\"\" class=\"k-textbox\" [(ngModel)]=\"item[valueField]\">\r\n                    </td>\r\n                    <td>\r\n                        <input type=\"text\" style=\"width:100%\" placeholder=\"\" autocomplete=\"off\" class=\"k-textbox\" [(ngModel)]=\"item[textField]\">\r\n                    </td>\r\n                </tr>\r\n            </tbody>\r\n        </table>\r\n    <!-- </perfect-scrollbar> -->\r\n</div>\r\n<div style=\"flex-shrink: 0; margin: 0 14px;\">\r\n<enum-editor-footer\r\n[showSortBtns]=\"showSortBtns\"\r\n(addFilter)=\"onAddFilter()\"\r\n(clearFilter)=\"onClear()\"\r\n(moveTop)=\"onMoveTop()\"\r\n(movePrev)=\"onMovePrev()\"\r\n(moveNext)=\"onMoveNext()\"\r\n(moveBottom)=\"onMoveBottom()\"></enum-editor-footer>\r\n</div>"
            }] }
];
/** @nocollapse */
EnumEditorComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: MessagerService },
    { type: LocaleService }
];
EnumEditorComponent.propDecorators = {
    data: [{ type: Input }],
    columns: [{ type: Input }],
    textField: [{ type: Input }],
    valueField: [{ type: Input }],
    dataChange: [{ type: Output }],
    tablebodybox: [{ type: ViewChild, args: ['tablebodybox',] }],
    cls: [{ type: HostBinding, args: ['class',] }],
    overflow: [{ type: HostBinding, args: ['style.overflow',] }],
    height: [{ type: HostBinding, args: ['style.height',] }],
    showSortBtns: [{ type: Input }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const EnumOutType = {
    'array': 0,
    'object': 1,
};
EnumOutType[EnumOutType['array']] = 'array';
EnumOutType[EnumOutType['object']] = 'object';
class EnumEditorService {
    /**
     * @param {?} injector
     * @param {?} cfr
     * @param {?} modalService
     * @param {?} localeService
     */
    constructor(injector, cfr, modalService, localeService) {
        this.injector = injector;
        this.cfr = cfr;
        this.modalService = modalService;
        this.localeService = localeService;
        this.enumEditorRef = null;
        this.dlgRef = null;
        this.originalData = [];
        this.textField = 'name';
        this.valueField = 'value';
        this.outType = EnumOutType.array;
        this.dataChanged = new Subject();
    }
    /**
     * @param {?=} enumData
     * @param {?=} options
     * @return {?}
     */
    showDialog(enumData = [], options = {}) {
        enumData = this.toJSON(enumData);
        /** @type {?} */
        const enumEditorFactory = this.cfr.resolveComponentFactory(EnumEditorComponent);
        this.enumEditorRef = enumEditorFactory.create(this.injector);
        this.enumEditorRef.instance.data = enumData;
        this.originalData = cloneDeep(enumData);
        this.textField = options['textField'] || 'name';
        this.valueField = options['valueField'] || 'value';
        this.outType = options['outType'] || EnumOutType.array;
        options['showSortBtns'] = this.outType === EnumOutType.array;
        this.dlgRef = this.modalService.show(this.enumEditorRef, {
            width: 800, height: 500,
            title: this.localeService.getValue('enumEditor.title'), enableScroll: false,
            minHeight: 398, minWidth: 798, iconCls: 'f-icon f-icon-top_developmenttool',
            buttons: [
                {
                    text: this.localeService.getValue('enumEditor.cancelButton'),
                    cls: 'btn btn-outline-secondary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        this.cancel();
                    })
                },
                {
                    text: this.localeService.getValue('enumEditor.okButton'),
                    cls: 'btn btn-primary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        this.save();
                    })
                }
            ],
            initialState: options,
            dialogFooterStyles: { background: '#F4F6F9' },
            buttonAlign: 'right',
            closed: (/**
             * @param {?} isCloseButtonClick
             * @return {?}
             */
            (isCloseButtonClick) => {
                if (isCloseButtonClick) {
                    this.cancel();
                }
            })
        });
        this.enumEditorRef.instance.height = this.dlgRef.dialog.instance.getContainerSize().height;
        this.dlgRef.dialog.instance.resized.subscribe((/**
         * @param {?} size
         * @return {?}
         */
        size => {
            this.enumEditorRef.instance.height = size.containerHeight;
        }));
        this.enumEditorRef.changeDetectorRef.detectChanges();
        this.dlgRef.dialog.changeDetectorRef.reattach();
        // this.openDialog.emit();
    }
    /**
     * @return {?}
     */
    cancel() {
        this._dataChanged(this.originalData);
        this.dlgRef.close();
    }
    /**
     * @return {?}
     */
    save() {
        /** @type {?} */
        const enumData = this.enumEditorRef.instance.data;
        this._dataChanged(enumData);
        this.dlgRef.close();
    }
    /**
     * @private
     * @param {?} enumData
     * @return {?}
     */
    _dataChanged(enumData) {
        /** @type {?} */
        const str = this.toString(enumData);
        this.dataChanged.next(str);
    }
    /**
     * @param {?} enumData
     * @return {?}
     */
    toString(enumData) {
        if (enumData && enumData.length) {
            if (this.outType === EnumOutType.array) {
                return JSON.stringify(enumData);
            }
            else {
                /** @type {?} */
                const obj = enumData.reduce((/**
                 * @param {?} r
                 * @param {?} c
                 * @return {?}
                 */
                (r, c) => {
                    r[c[this.valueField]] = c[this.textField];
                    return r;
                }), {});
                return JSON.stringify(obj);
            }
        }
        return '';
    }
    /**
     * @param {?} val
     * @return {?}
     */
    toJSON(val) {
        /** @type {?} */
        let enumData = [];
        if (val) {
            if (typeof val === 'string') {
                try {
                    enumData = JSON.parse(val);
                    if (this.outType === EnumOutType.object) {
                        enumData = this.convertObject2Array(enumData);
                    }
                }
                catch (e) {
                    enumData = [];
                    // console.warn(e);
                }
            }
            else {
                if (Array.isArray(val)) {
                    enumData = val;
                }
                else {
                    enumData = this.convertObject2Array(val);
                }
            }
        }
        return enumData;
    }
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    convertObject2Array(obj) {
        /** @type {?} */
        const _enumData = [];
        Object.keys(obj).forEach((/**
         * @param {?} n
         * @return {?}
         */
        n => {
            _enumData.push({ [this.valueField]: n, [this.textField]: obj[n] });
        }));
        return _enumData;
    }
}
EnumEditorService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
EnumEditorService.ctorParameters = () => [
    { type: Injector },
    { type: ComponentFactoryResolver },
    { type: BsModalService },
    { type: LocaleService }
];
/** @nocollapse */ EnumEditorService.ngInjectableDef = defineInjectable({ factory: function EnumEditorService_Factory() { return new EnumEditorService(inject(INJECTOR), inject(ComponentFactoryResolver), inject(BsModalService), inject(LocaleService)); }, token: EnumEditorService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class EnumEditorFooterComponent {
    /**
     * @param {?} grid
     */
    constructor(grid) {
        this.grid = grid;
        this.cls = 'f-filter-footer';
        this.showSortBtns = true;
        this.addFilter = new EventEmitter();
        this.clearFilter = new EventEmitter();
        this.moveTop = new EventEmitter();
        this.movePrev = new EventEmitter();
        this.moveNext = new EventEmitter();
        this.moveBottom = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    onAddFilter() {
        this.addFilter.emit();
    }
    /**
     * @return {?}
     */
    onClear() {
        this.clearFilter.emit();
    }
    /**
     * @return {?}
     */
    onMoveTop() {
        this.moveTop.emit();
    }
    /**
     * @return {?}
     */
    onMovePrev() {
        this.movePrev.emit();
    }
    /**
     * @return {?}
     */
    onMoveNext() {
        this.moveNext.emit();
    }
    /**
     * @return {?}
     */
    onMoveBottom() {
        this.moveBottom.emit();
    }
    /**
     * @return {?}
     */
    canUse() {
        if (this.grid.currentEnumData) {
            return this.grid.data.length > 1;
        }
        return false;
    }
    /**
     * @return {?}
     */
    canMoveUp() {
        if (this.canUse()) {
            return this.grid.currentEnumData.index > 0;
        }
        return false;
    }
    /**
     * @return {?}
     */
    canMoveDown() {
        if (this.canUse()) {
            return this.grid.currentEnumData.index < this.grid.data.length - 1;
        }
        return false;
    }
}
EnumEditorFooterComponent.decorators = [
    { type: Component, args: [{
                selector: 'enum-editor-footer',
                template: `
    <div class="footer-container px-2" style="height: 36px;">
        <button type="button" (click)="onAddFilter()"
            class="k-button k-button-icontext k-flat "><span class="k-icon k-i-add"></span> {{ 'enumEditor.addWhere' | locale }}</button>
        <button type="button" (click)="onClear()" [disabled]="!grid.data.length"
            class="k-button k-button-icontext k-flat "><span class="k-icon k-i-delete"></span> {{ 'enumEditor.clear' | locale }}</button>
        <ng-container *ngIf="showSortBtns">
            <button type="button" [disabled]="!canMoveUp()" (click)="onMoveTop()"
            class="k-button k-button-icontext k-flat "><span class="k-icon k-i-arrow-end-up"></span>
            {{'enumEditor.moveTop' | locale }}</button>
        <button type="button" [disabled]="!canMoveUp()" (click)="onMovePrev()"
            class="k-button k-button-icontext k-flat "><span class="k-icon k-i-arrow-60-up"></span>
            {{'enumEditor.moveUp' | locale }}</button>
        <button type="button" [disabled]="!canMoveDown()" (click)="onMoveNext()"
            class="k-button k-button-icontext k-flat "><span class="k-icon k-i-arrow-60-down"></span>
            {{'enumEditor.moveDown' | locale }}</button>
        <button type="button" [disabled]="!canMoveDown()" (click)="onMoveBottom()"
            class="k-button k-button-icontext k-flat"><span class="k-icon k-i-arrow-end-down"></span>
            {{'enumEditor.moveBottom' | locale }}</button>
            </ng-container>
    </div>
    `
            }] }
];
/** @nocollapse */
EnumEditorFooterComponent.ctorParameters = () => [
    { type: EnumEditorComponent }
];
EnumEditorFooterComponent.propDecorators = {
    cls: [{ type: HostBinding, args: ['class',] }],
    showSortBtns: [{ type: Input }],
    addFilter: [{ type: Output }],
    clearFilter: [{ type: Output }],
    moveTop: [{ type: Output }],
    movePrev: [{ type: Output }],
    moveNext: [{ type: Output }],
    moveBottom: [{ type: Output }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const ENUM_EDITOR_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef((/**
     * @return {?}
     */
    () => EnumEditorInputComponent)),
    multi: true
};
class EnumEditorInputComponent extends LookupComponent {
    /**
     * @param {?} injector
     * @param {?} ees
     * @param {?} el
     */
    constructor(injector, ees, el) {
        super(injector, el);
        this.injector = injector;
        this.ees = ees;
        this.el = el;
        this.textField = 'name';
        this.valueField = 'value';
        this.outType = EnumOutType.array;
        this.enableClear = true;
        this.valueChange = new EventEmitter();
        this.groupIcon = '<i class="f-icon f-icon-lookup"></i>';
        this.cd = null;
        this.ees.dataChanged.subscribe((/**
         * @param {?} str
         * @return {?}
         */
        str => {
            this.updateData(str);
        }));
        this.cd = this.injector.get(ChangeDetectorRef, null);
    }
    /**
     * @private
     * @param {?} str
     * @return {?}
     */
    updateData(str) {
        this.value = str;
        if (this.cd) {
            this.cd.detectChanges();
        }
        this.valueChange.emit(str);
        this.onModelChange(str);
        this.onModelTouched(str);
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    showDialog() {
        /** @type {?} */
        const enumData = this.convertValue2JSON();
        this.ees.showDialog(enumData, {
            textField: this.textField,
            valueField: this.valueField,
            outType: this.outType
        });
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClear($event) {
        this.updateData('');
    }
    /**
     * @param {?} val
     * @return {?}
     */
    writeValue(val) {
        if (val && typeof val === 'object') {
            this.value = JSON.stringify(val);
        }
        else {
            this.value = val || '';
        }
    }
    /**
     * @private
     * @return {?}
     */
    convertValue2JSON() {
        return this.ees.toJSON(this.value);
    }
}
EnumEditorInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'enum-editor-input',
                template: `
    <input-group #ig style="width: 100%"
        [groupText]="groupIcon"
        (clear)="onClear($event)"
        (clickHandle)="showDialog()"
        [readonly] = "readonly"
        [editable] = "editable"
        [enableClear] = "enableClear"
        [(value)]="value">
    </input-group>
    `,
                providers: [
                    ENUM_EDITOR_VALUE_ACCESSOR,
                    EnumEditorService
                ]
            }] }
];
/** @nocollapse */
EnumEditorInputComponent.ctorParameters = () => [
    { type: Injector },
    { type: EnumEditorService },
    { type: ElementRef }
];
EnumEditorInputComponent.propDecorators = {
    value: [{ type: Input }],
    textField: [{ type: Input }],
    valueField: [{ type: Input }],
    outType: [{ type: Input }],
    enableClear: [{ type: Input }],
    valueChange: [{ type: Output }],
    inputGroupRef: [{ type: ViewChild, args: ['ig',] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class EnumEditorModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: EnumEditorModule,
            providers: [
                EnumEditorService
            ]
        };
    }
}
EnumEditorModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    EnumEditorComponent,
                    EnumEditorInputComponent,
                    EnumEditorFooterComponent
                ],
                imports: [
                    CommonModule,
                    FormsModule,
                    InputGroupModule,
                    FarrisCommonModule,
                    MessagerModule.forRoot(),
                    PerfectScrollbarModule,
                    LookupModule,
                    LocaleModule
                ],
                entryComponents: [
                    EnumEditorComponent
                ],
                exports: [EnumEditorComponent, EnumEditorInputComponent]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { EnumOutType, EnumEditorService, EnumEditorComponent, EnumEditorFooterComponent, ENUM_EDITOR_VALUE_ACCESSOR, EnumEditorInputComponent, EnumEditorModule };
