// @farris/devkit/esm5/lib/frame/component_ref_manager.js
/**
 * 组件实例管理
 */
var ComponentRefManager = /** @class */ (function () {
    function ComponentRefManager() {
        this.components = new Map();
    }
    /**
     * 增加组件实例
     * @param id
     * @param component
     */
    ComponentRefManager.prototype.add = function (id, component) {
        var components = this.components.get(id);
        if (components) {
            components.push(component);
        }
        else {
            this.components.set(id, [component]);
        }
    };
    /**
     * 删除组件实例
     * @param id
     */
    ComponentRefManager.prototype.remove = function (id) {
        var components = this.components.get(id);
        if (components && components.length > 0) {
            components.length = 0;
            this.components.delete(id);
        }
    };
    /**
     * 通过id获取组件实例
     * @param id
     * @returns
     */
    ComponentRefManager.prototype.get = function (id) {
        if (id) {
            var components = this.components.get(id);
            if (components && components.length > 0) {
                return components;
            }
        }
        return null;
    };
    /**
     * 是否包含id的组件
     * @param id
     * @returns
     */
    ComponentRefManager.prototype.has = function (id) {
        return this.components.has(id);
    };
    /**
     * 清空
     */
    ComponentRefManager.prototype.clear = function () {
        this.components.clear();
    };
    /**
     * 获取指定类型的组件
     * @param componentType
     * @returns
     */
    ComponentRefManager.prototype.getComponentByType = function (componentType) {
        var containers = Array.from(this.components.values());
        var component = null;
        for (var containerIndex = 0; containers && containerIndex < containers.length; containerIndex++) {
            var components = containers[containerIndex];
            for (var index = 0; components && index < components.length; index++) {
                if (components[index] instanceof componentType) {
                    component = components[index];
                    break;
                }
            }
        }
        return component;
    };
    return ComponentRefManager;
}());
export { ComponentRefManager };
