// @farris/devkit/esm5/lib/frame/frame_event_bus.js
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { filter } from 'rxjs/operators';
/**
 * 框架事件总线
 */
var FrameEventBus = /** @class */ (function () {
    function FrameEventBus() {
        /**
         * 事件流
         */
        this.events = new Subject();
    }
    /**
     * 注册事件处理
     */
    FrameEventBus.prototype.on = function (eventType, eventHandler, frameId) {
        this.events.pipe(filter(function (event) {
            if (event.type !== eventType) {
                return false;
            }
            return !event.frameIds || event.frameIds.indexOf(frameId) > -1;
        })).subscribe(eventHandler);
    };
    /**
     * 取消事件处理
     */
    FrameEventBus.prototype.off = function (eventType, frameId) {
        throw new Error('暂不实现');
    };
    /**
     * 触发事件
     */
    FrameEventBus.prototype.trigger = function (eventType, eventData, frameIds) {
        var event = {
            type: eventType,
            data: eventData,
            frameIds: frameIds
        };
        this.events.next(event);
    };
    FrameEventBus.decorators = [
        { type: Injectable }
    ];
    return FrameEventBus;
}());
export { FrameEventBus };
