// @farris/kendo-binding/esm2015/lib/pipes/formatPrimeNgTreeTableCellData.js
import { LocaleService } from '@farris/ui-locale';
import { Pipe, Injector, Optional } from '@angular/core';
import { ColumnFormatService } from '@farris/ui-common/column';
export class FormatPrimeNgTreeTableCellData {
    constructor(cfs, injector) {
        this.cfs = cfs;
        this.injector = injector;
        if (this.injector) {
            this.localeService = this.injector.get(LocaleService);
        }
    }
    transform(col, rowData) {
        let value = '';
        if (rowData && col && col.field) {
            value = this.getValue(col.field, rowData);
            let formatterFn = col.formatter;
            if (!formatterFn) {
                if (col.isMultilingualField) {
                    return this.getMultiLanguageValue(value);
                }
                else {
                    return value;
                }
            }
            else {
                return this.cfs.format(value, rowData, formatterFn);
            }
        }
        return '';
    }
    // 获取多语数据
    getMultiLanguageValue(valObj) {
        if (valObj && typeof valObj === 'object' && Object.keys(valObj).length > 0) {
            const localeId = this.localeService && this.localeService.localeId || 'zh-CHS';
            const value = valObj[localeId];
            return value === undefined ? '' : value;
        }
        else {
            return '';
        }
    }
    getValue(path, target) {
        if (!target) {
            return '';
        }
        var resultVal = '';
        if (path.indexOf('.') === -1) {
            resultVal = target[path];
        }
        else {
            resultVal = path.split('.').reduce(((obj, key) => {
                return obj && obj[key] || null;
            }), target);
        }
        return resultVal;
    }
    ;
    /**
     * 是否为帮助列
     * @param col col
     */
    isHelpColumn(col) {
        return col && col.editor && col.editor.options && col.editor.options.type === 'EditorTypes.LOOKUP' || false;
    }
    getHelpColumnValue(col, data) {
        const textField = this.getHelpControlTextField(col);
        const mapFields = this.getHelpControlMapFields(col);
        if (textField && mapFields && Object.keys(mapFields).length > 0) {
            const path = mapFields[textField];
            return this.getValue(path, data);
        }
        return '';
    }
    getHelpControlTextField(col) {
        return col && col.editor && col.editor.options && col.editor.options.textField || '';
    }
    getHelpControlMapFields(col) {
        return col && col.editor && col.editor.options && col.editor.options.mapFields || {};
    }
}
FormatPrimeNgTreeTableCellData.decorators = [
    { type: Pipe, args: [{ name: 'formatPrimeNgTreeTableCellData', pure: false },] }
];
/** @nocollapse */
FormatPrimeNgTreeTableCellData.ctorParameters = () => [
    { type: ColumnFormatService },
    { type: Injector, decorators: [{ type: Optional }] }
];
