// @gsp-cmp/chgdr/bundles/gsp-cmp-chgdr.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/forms'), require('@farris/ui-splitter'), require('@farris/ui-draggable'), require('@farris/ui-datagrid'), require('@farris/ui-treetable'), require('@farris/ui-section'), require('@farris/ui-dialog'), require('@farris/ui-datepicker'), require('@ngx-translate/core'), require('@gsp-lcm/metadatart-selector'), require('date-fns'), require('@gsp-bef/gsp-cm-metadata'), require('@angular/common/locales/zh'), require('@ecp-caf/caf-common'), require('@farris/ui-modal'), require('@angular/core'), require('@farris/ui-messager'), require('@farris/ui-notify'), require('rxjs/operators'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@gsp-cmp/chgdr', ['exports', '@angular/common', '@angular/forms', '@farris/ui-splitter', '@farris/ui-draggable', '@farris/ui-datagrid', '@farris/ui-treetable', '@farris/ui-section', '@farris/ui-dialog', '@farris/ui-datepicker', '@ngx-translate/core', '@gsp-lcm/metadatart-selector', 'date-fns', '@gsp-bef/gsp-cm-metadata', '@angular/common/locales/zh', '@ecp-caf/caf-common', '@farris/ui-modal', '@angular/core', '@farris/ui-messager', '@farris/ui-notify', 'rxjs/operators', 'rxjs'], factory) :
    (factory((global['gsp-cmp'] = global['gsp-cmp'] || {}, global['gsp-cmp'].chgdr = {}),global.ng.common,global.ng.forms,global.uiSplitter,global.uiDraggable,global.uiDatagrid,global.uiTreetable,global.uiSection,global.uiDialog,global.uiDatepicker,global.core,global.i2,global.dateFns,global.gspCmMetadata,global.ng.common.locales.zh,global.i1,global.i1$1,global.ng.core,global.i2$1,global.uiNotify,global.rxjs.operators,global.rxjs));
}(this, (function (exports,common,forms,uiSplitter,uiDraggable,uiDatagrid,uiTreetable,uiSection,uiDialog,uiDatepicker,core,i2,dateFns,gspCmMetadata,zh,i1,i1$1,i0,i2$1,uiNotify,operators,rxjs) { 'use strict';

    zh = zh && zh.hasOwnProperty('default') ? zh['default'] : zh;

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/chgdr-data-viewer/chgdr-data.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 基本信息
     */
    var /**
     * 基本信息
     */ ChgdrData = /** @class */ (function () {
        function ChgdrData() {
        }
        return ChgdrData;
    }());
    /**
     * 发生变更的实体节点
     */
    var /**
     * 发生变更的实体节点
     */ ChgdrDataEntityNode = /** @class */ (function () {
        function ChgdrDataEntityNode() {
        }
        return ChgdrDataEntityNode;
    }());
    /**
     * 发生变更的行
     */
    var /**
     * 发生变更的行
     */ ChgdrDataRowNode = /** @class */ (function () {
        function ChgdrDataRowNode() {
        }
        return ChgdrDataRowNode;
    }());
    var ChangeColumnItem = /** @class */ (function () {
        function ChangeColumnItem() {
            /**
             * 如果当前字段为关联字段，则属性为当前字段的关联带出字段
             */
            this.children = [];
        }
        return ChangeColumnItem;
    }());
    var DataCodeField = /** @class */ (function () {
        function DataCodeField() {
        }
        return DataCodeField;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/service/i18n/en.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var LangDataEn = {
        "chgdr-list-dialog-title": "Change data records",
        "base-info-section-title": "Base info",
        "change-detail-section-title": "Change detail",
        "empty-data-text": "Empty Data",
        "get-business-code-config-error": "Get business code config error: {{errorMessage}}",
        "user-name-column-title": "User",
        "change-time-column-title": "Time",
        "change-time-column-format": "yyyy-MM-dd HH:mm:ss",
        "operation-type-title": "Type",
        "tech-pk-column-title": "Tech PK",
        "operation-column-title": "Operation",
        "can-not-found-change-record-by-id": "Can not found change record 【{{headerId}}】",
        "query-change-data-record-error": "Query change data record error: {{errorMessage}}",
        "show-change-detail-btn": "Detail",
        "change-detail-dialog-title": "Change detail",
        "start-time-label": "Start",
        "end-time-label": "End",
        "query-btn": "Query",
        "technology-primary-key-label": "Tech PK",
        "change-user-label": "User",
        "change-time-label": "Time",
        "change-operate-type": "Type",
        "field-name-column-title": "Field",
        "value-column-title": "Value",
        "old-value-column-title": "Old value",
        "new-value-column-title": "New value",
        "operate-type-name-ADD": "Add",
        "operate-type-name-MODIFY": "Modify",
        "operate-type-name-DELETE": "Delete",
        "yes-text": "Yes",
        "no-text": "No",
        "business-code-text": "Business code"
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/service/i18n/zh-CHS.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var LangDataZhChs = {
        "chgdr-list-dialog-title": "变更日志列表",
        "base-info-section-title": "基本信息",
        "change-detail-section-title": "变更详情",
        "empty-data-text": "无数据",
        "get-business-code-config-error": "获取业务编号配置失败：{{errorMessage}}",
        "user-name-column-title": "用户",
        "change-time-column-title": "时间",
        "change-time-column-format": "yyyy年MM月dd日 HH:mm:ss",
        "operation-type-title": "操作类型",
        "tech-pk-column-title": "技术主键",
        "operation-column-title": "操作",
        "can-not-found-change-record-by-id": "未找到编号为【{{headerId}}】的变更日志",
        "query-change-data-record-error": "查询业务变更日志出错：{{errorMessage}}",
        "show-change-detail-btn": "变更详情",
        "change-detail-dialog-title": "变更日志详情",
        "start-time-label": "开始时间",
        "end-time-label": "结束时间",
        "query-btn": "查询",
        "technology-primary-key-label": "技术主键",
        "change-user-label": "变更人",
        "change-time-label": "变更时间",
        "change-operate-type": "变更方式",
        "field-name-column-title": "字段名",
        "value-column-title": "值",
        "old-value-column-title": "原值",
        "new-value-column-title": "变更值",
        "operate-type-name-ADD": "新增",
        "operate-type-name-MODIFY": "修改",
        "operate-type-name-DELETE": "删除",
        "yes-text": "是",
        "no-text": "否",
        "business-code-text": "业务编号"
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/service/i18n/zh-CHT.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var LangDataZhCht = {
        "chgdr-list-dialog-title": "變更日志列表",
        "base-info-section-title": "基本信息",
        "change-detail-section-title": "變更詳情",
        "empty-data-text": "無數據",
        "get-business-code-config-error": "獲取業務編號配置失敗：{{errorMessage}}",
        "user-name-column-title": "用戶",
        "change-time-column-title": "時間",
        "change-time-column-format": "yyyy年MM月dd日 HH:mm:ss",
        "operation-type-title": "操作類型",
        "tech-pk-column-title": "技術主鍵",
        "operation-column-title": "操作",
        "can-not-found-change-record-by-id": "未找到編號爲【{{headerId}}】的變更日志",
        "query-change-data-record-error": "查詢業務變更日志出錯：{{errorMessage}}",
        "show-change-detail-btn": "變更詳情",
        "change-detail-dialog-title": "變更日志詳情",
        "start-time-label": "開始時間",
        "end-time-label": "結束時間",
        "query-btn": "查詢",
        "technology-primary-key-label": "技術主鍵",
        "change-user-label": "變更人",
        "change-time-label": "變更時間",
        "change-operate-type": "變更方式",
        "field-name-column-title": "字段名",
        "value-column-title": "值",
        "old-value-column-title": "原值",
        "new-value-column-title": "變更值",
        "operate-type-name-ADD": "新增",
        "operate-type-name-MODIFY": "修改",
        "operate-type-name-DELETE": "刪除",
        "yes-text": "是",
        "no-text": "否",
        "business-code-text": "業務編號"
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/service/i18n-service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var langResourceMap = new Map();
    langResourceMap.set("en", LangDataEn);
    langResourceMap.set("zh-CHS", LangDataZhChs);
    langResourceMap.set("zh-CHT", LangDataZhCht);
    /** @type {?} */
    var currentLang = localStorage.getItem('languageCode');
    if (!currentLang) {
        currentLang = "zh-CHS";
        console.log("未找到设置的语言，默认使用" + currentLang);
    }
    /** @type {?} */
    var parser = new core.TranslateDefaultParser();
    var I18nUtil = /** @class */ (function () {
        function I18nUtil() {
        }
        /**
         * @param {?} key
         * @param {?=} param
         * @return {?}
         */
        I18nUtil.instant = /**
         * @param {?} key
         * @param {?=} param
         * @return {?}
         */
            function (key, param) {
                /** @type {?} */
                var statement = I18nUtil.getCurrentLangResource()[key];
                if (statement === undefined || statement === null) {
                    return key;
                }
                /** @type {?} */
                var result = parser.interpolate(statement, param);
                if (result === undefined || result === null) {
                    return key;
                }
                return result;
            };
        /**
         * @private
         * @return {?}
         */
        I18nUtil.getCurrentLangResource = /**
         * @private
         * @return {?}
         */
            function () {
                return langResourceMap.get(currentLang) || {};
            };
        /**
         * @param {?} key
         * @param {?} defaultText
         * @return {?}
         */
        I18nUtil.instantOrDefault = /**
         * @param {?} key
         * @param {?} defaultText
         * @return {?}
         */
            function (key, defaultText) {
                /** @type {?} */
                var statement = I18nUtil.getCurrentLangResource()[key];
                if (statement === undefined || statement === null) {
                    return defaultText;
                }
                /** @type {?} */
                var result = parser.interpolate(statement);
                if (result === undefined || result === null) {
                    return defaultText;
                }
                return result;
            };
        return I18nUtil;
    }());
    var I18nService = /** @class */ (function () {
        function I18nService() {
        }
        /**
         * @param {?} key
         * @param {?=} param
         * @return {?}
         */
        I18nService.prototype.instant = /**
         * @param {?} key
         * @param {?=} param
         * @return {?}
         */
            function (key, param) {
                return I18nUtil.instant(key, param);
            };
        /**
         * @private
         * @return {?}
         */
        I18nService.prototype.getCurrentLangResource = /**
         * @private
         * @return {?}
         */
            function () {
                return langResourceMap.get(currentLang) || {};
            };
        /**
         * @param {?} key
         * @param {?} defaultText
         * @return {?}
         */
        I18nService.staticInstant = /**
         * @param {?} key
         * @param {?} defaultText
         * @return {?}
         */
            function (key, defaultText) {
                return I18nUtil.instantOrDefault(key, defaultText);
            };
        I18nService.decorators = [
            { type: i0.Injectable, args: [{ providedIn: 'root' },] }
        ];
        /** @nocollapse */
        I18nService.ctorParameters = function () { return []; };
        /** @nocollapse */ I18nService.ngInjectableDef = i0.defineInjectable({ factory: function I18nService_Factory() { return new I18nService(); }, token: I18nService, providedIn: "root" });
        return I18nService;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/chgdr-data-viewer/chgdr-data-builder.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChgdrDataBuilder = /** @class */ (function () {
        function ChgdrDataBuilder() {
        }
        /**
         * @param {?} data
         * @param {?} handler
         * @return {?}
         */
        ChgdrDataBuilder.prototype.buildChgdrData = /**
         * @param {?} data
         * @param {?} handler
         * @return {?}
         */
            function (data, handler) {
                var _this = this;
                this.handler = handler;
                //初始化变更日志基本信息
                /** @type {?} */
                var chgdrData = new ChgdrData();
                chgdrData.entityName = this.handler.getBeName();
                chgdrData.operateType = data.operateType;
                chgdrData.userName = data.userName;
                chgdrData.dataId = data.dataId;
                chgdrData.dataCode = data.dataCode;
                chgdrData.dataCodes = this.buildDataCodes(data.dataCode, this.handler.getMainObjectCode());
                chgdrData.reason = data.reason;
                chgdrData.changeTime = data.changeTime;
                chgdrData.entityNodes = [];
                //分层组织变更日志行
                /** @type {?} */
                var rowNodes = [];
                !!data.rows && data.rows.forEach(( /**
                 * @param {?} row
                 * @return {?}
                 */function (row) {
                    if (!row) {
                        return;
                    }
                    /** @type {?} */
                    var rowNode = new ChgdrDataRowNode();
                    rowNode.id = row.id;
                    rowNode.parentDataId = row.parentDataId;
                    rowNode.dataId = row.dataId;
                    rowNode.dataCode = row.dataCode;
                    rowNode.dataCodes = _this.buildDataCodes(row.dataCode, row.entityCode);
                    rowNode.entityCode = row.entityCode;
                    rowNode.entityName = _this.handler.getEntityNameOrDefault(row.entityCode);
                    rowNode.operateType = row.operateType;
                    rowNode.entityNodes = [];
                    //组织列的变更信息
                    /** @type {?} */
                    var changeColumnKeys = _this.extractChangeColumnKeys(row);
                    /** @type {?} */
                    var columnValueChanges = [];
                    changeColumnKeys.forEach(( /**
                     * @param {?} key
                     * @return {?}
                     */function (key) {
                        /** @type {?} */
                        var column = new ChangeColumnItem();
                        column.fieldLabel = key;
                        column.fieldName = _this.handler.getEntityFieldNameOrDefault(row.entityCode, key);
                        column.oldValue = _this.handler.formatFieldValue(row.oldContent.get(key), row.entityCode, key);
                        column.newValue = _this.handler.formatFieldValue(row.newContent.get(key), row.entityCode, key);
                        /** @type {?} */
                        var ele = _this.handler.getElement(row.entityCode, key);
                        if (ele) {
                            //判断是否是关联字段
                            column.hasAssociation = ele.ObjectType == "Association" && ele.IsUdt == false;
                            column.isAssociationRefField = ele.IsRefElement;
                            /** @type {?} */
                            var parentElement = _this.handler.getParentElement(row.entityCode, key);
                            if (parentElement) {
                                column.belongFieldLabelId = parentElement.LabelID;
                            }
                        }
                        columnValueChanges.push(column);
                    }));
                    //变更值排序
                    _this.handler.sort(columnValueChanges, row.entityCode, ( /**
                     * @param {?} column
                     * @return {?}
                     */function (column) { return column.fieldLabel; }));
                    //合并关联带出字段
                    /** @type {?} */
                    var columnMap = new Map();
                    columnValueChanges.forEach(( /**
                     * @param {?} column
                     * @return {?}
                     */function (column) {
                        columnMap.set(column.fieldLabel, column);
                    }));
                    //1. 移除所有关联带出字段，并附加到其所属关联字段的children属性上
                    columnValueChanges = columnValueChanges.filter(( /**
                     * @param {?} column
                     * @return {?}
                     */function (column) {
                        if (column.isAssociationRefField) {
                            /** @type {?} */
                            var parentColumn = columnMap.get(column.belongFieldLabelId);
                            if (parentColumn) {
                                //找到父级字段则添加为父
                                parentColumn.children.push(column);
                                return false;
                            }
                            else {
                                return true;
                            }
                        }
                        else {
                            return true;
                        }
                    }));
                    //2. 转换关联字段，将其子级的变更值拼接作为其变更值
                    columnValueChanges = columnValueChanges.map(( /**
                     * @param {?} column
                     * @return {?}
                     */function (column) {
                        if (column.hasAssociation && column.children.length > 0) {
                            //如果是关联字段，且存在关联带出字段
                            /** @type {?} */
                            var virtualColumn = Object.assign(new ChangeColumnItem(), column);
                            virtualColumn.originalColumnItem = column;
                            virtualColumn.oldValue = column.children.map(( /**
                             * @param {?} child
                             * @return {?}
                             */function (child) { return child.oldValue; })).join(";");
                            virtualColumn.newValue = column.children.map(( /**
                             * @param {?} child
                             * @return {?}
                             */function (child) { return child.newValue; })).join(";");
                            return virtualColumn;
                        }
                        else {
                            return column;
                        }
                    }));
                    rowNode.changes = columnValueChanges;
                    rowNodes.push(rowNode);
                }));
                /**
                 * dataId与变更行的Map
                 * @type {?}
                 */
                var dataIdRowMap = new Map();
                rowNodes.forEach(( /**
                 * @param {?} node
                 * @return {?}
                 */function (node) { return dataIdRowMap.set(node.dataId, node); }));
                rowNodes.forEach(( /**
                 * @param {?} node
                 * @return {?}
                 */function (node) {
                    if (!node.parentDataId) {
                        //根节点
                        /** @type {?} */
                        var entityNode = _this.genChgdrDataEntityNode(node.entityCode, node.entityName);
                        entityNode.rows.push(node);
                        chgdrData.entityNodes.push(entityNode);
                        return;
                    }
                    /** @type {?} */
                    var parentRowNode = dataIdRowMap.get(node.parentDataId);
                    if (!!parentRowNode) {
                        //上级节点存在的节点
                        /** @type {?} */
                        var entityNode = parentRowNode.entityNodes.find(( /**
                         * @param {?} en
                         * @return {?}
                         */function (en) { return en.entityCode == node.entityCode; }));
                        if (!entityNode) {
                            entityNode = _this.genChgdrDataEntityNode(node.entityCode, node.entityName);
                            parentRowNode.entityNodes.push(entityNode);
                        }
                        entityNode.rows.push(node);
                    }
                    else {
                        //上级节点不存在的节点
                        //TODO 上级节点不存在的节点是否需要构造完整的上级结构？
                        /** @type {?} */
                        var entityNode = _this.genChgdrDataEntityNode(node.entityCode, node.entityName);
                        entityNode.rows.push(node);
                        chgdrData.entityNodes.push(entityNode);
                    }
                }));
                return chgdrData;
            };
        /**
         * @private
         * @param {?} dataCode
         * @param {?} entityCode
         * @return {?}
         */
        ChgdrDataBuilder.prototype.buildDataCodes = /**
         * @private
         * @param {?} dataCode
         * @param {?} entityCode
         * @return {?}
         */
            function (dataCode, entityCode) {
                var _this = this;
                /** @type {?} */
                var dataCodes = [];
                dataCode && dataCode.forEach(( /**
                 * @param {?} value
                 * @param {?} key
                 * @return {?}
                 */function (value, key) {
                    /** @type {?} */
                    var dataCodeField = new DataCodeField();
                    if (key == "_$dataCode") {
                        dataCodeField.fieldLabelId = "dataCode";
                        dataCodeField.fieldName = I18nUtil.instantOrDefault("business-code-text", "业务编号");
                        dataCodeField.fieldValue = value;
                    }
                    else {
                        dataCodeField.fieldLabelId = key;
                        dataCodeField.fieldName = _this.handler.getEntityFieldNameOrDefault(entityCode, key);
                        dataCodeField.fieldValue = _this.handler.formatFieldValue(value, entityCode, key);
                    }
                    dataCodes.push(dataCodeField);
                }));
                //业务编号排序
                this.handler.sort(dataCodes, entityCode, ( /**
                 * @param {?} dataCode
                 * @return {?}
                 */function (dataCode) { return dataCode.fieldLabelId; }));
                return dataCodes;
            };
        /** 得到新旧内容中所有的key */
        /**
         * 得到新旧内容中所有的key
         * @private
         * @param {?} row
         * @return {?}
         */
        ChgdrDataBuilder.prototype.extractChangeColumnKeys = /**
         * 得到新旧内容中所有的key
         * @private
         * @param {?} row
         * @return {?}
         */
            function (row) {
                /** @type {?} */
                var keysSet = new Set();
                row.oldContent && row.oldContent.forEach(( /**
                 * @param {?} value
                 * @param {?} key
                 * @return {?}
                 */function (value, key) { return keysSet.add(key); }));
                row.newContent && row.newContent.forEach(( /**
                 * @param {?} value
                 * @param {?} key
                 * @return {?}
                 */function (value, key) { return keysSet.add(key); }));
                /** @type {?} */
                var keys = Array.from(keysSet.keys());
                return keys;
            };
        /**
         * @private
         * @param {?} entityCode
         * @param {?} entityName
         * @return {?}
         */
        ChgdrDataBuilder.prototype.genChgdrDataEntityNode = /**
         * @private
         * @param {?} entityCode
         * @param {?} entityName
         * @return {?}
         */
            function (entityCode, entityName) {
                /** @type {?} */
                var node = new ChgdrDataEntityNode();
                node.entityCode = entityCode;
                node.entityName = entityName;
                node.rows = [];
                return node;
            };
        return ChgdrDataBuilder;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/util/guid.util.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var Guid = /** @class */ (function () {
        function Guid() {
        }
        /**
         * @return {?}
         */
        Guid.newGuid = /**
         * @return {?}
         */
            function () {
                var _this = this;
                /** @type {?} */
                var initString = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx';
                /** @type {?} */
                var result = initString.replace(/[xy]/g, ( /**
                 * @param {?} c
                 * @return {?}
                 */function (c) {
                    return _this.convert(c);
                }));
                return result;
            };
        /**
         * @private
         * @param {?} c
         * @return {?}
         */
        Guid.convert = /**
         * @private
         * @param {?} c
         * @return {?}
         */
            function (c) {
                // tslint:disable-next-line:no-bitwise
                /** @type {?} */
                var r = Math.random() * 16 | 0;
                // tslint:disable-next-line:no-bitwise
                /** @type {?} */
                var v = c === 'x' ? r : (r & 0x3 | 0x8);
                return v.toString(16);
            };
        return Guid;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/chgdr-data-viewer/chgdr-tree-builder.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    // 参考： Farris TreeTable的节点结构
    // export interface TreeNode {
    //     id?: string;
    //     data: any;
    //     /** 叶子节点图标 */
    //     icon?: any;
    //     /** 节点展开图标 */
    //     expandedIcon?: any;
    //     /** 节点折叠图标 */
    //     collapsedIcon?: any;
    //     /** 子节点 */
    //     children?: TreeNode[];
    //     /** 是否为叶子节点 */
    //     leaf?: boolean;
    //     /** 是否展开 */
    //     expanded?: boolean;
    //     type?: string;
    //     /** 父节点 */
    //     parent?: TreeNode;
    //     /** 是否允许选中 */
    //     selectable?: boolean;
    //  }
    var 
    // 参考： Farris TreeTable的节点结构
    // export interface TreeNode {
    //     id?: string;
    //     data: any;
    //     /** 叶子节点图标 */
    //     icon?: any;
    //     /** 节点展开图标 */
    //     expandedIcon?: any;
    //     /** 节点折叠图标 */
    //     collapsedIcon?: any;
    //     /** 子节点 */
    //     children?: TreeNode[];
    //     /** 是否为叶子节点 */
    //     leaf?: boolean;
    //     /** 是否展开 */
    //     expanded?: boolean;
    //     type?: string;
    //     /** 父节点 */
    //     parent?: TreeNode;
    //     /** 是否允许选中 */
    //     selectable?: boolean;
    //  }
    ChgdrTreeNode = /** @class */ (function () {
        function ChgdrTreeNode() {
        }
        return ChgdrTreeNode;
    }());
    var ChgdrTreeBuilder = /** @class */ (function () {
        function ChgdrTreeBuilder() {
        }
        /**
         * @param {?} chgdrData
         * @return {?}
         */
        ChgdrTreeBuilder.prototype.build = /**
         * @param {?} chgdrData
         * @return {?}
         */
            function (chgdrData) {
                this.chgdrData = chgdrData;
                if (chgdrData.entityNodes.length == 0) {
                    console.warn("无变更日志");
                    return [];
                }
                /** @type {?} */
                var rootEntityNode = chgdrData.entityNodes[0];
                /** @type {?} */
                var rootRows = rootEntityNode.rows;
                if (rootRows.length == 0) {
                    console.warn("无根变更日志行");
                    return [];
                }
                /** @type {?} */
                var rootRow = rootRows[0];
                /** @type {?} */
                var node = this.rowToNode(rootRow);
                node.isRoot = true;
                node.expanded = true;
                return [node];
            };
        /**
         * @private
         * @param {?} row
         * @return {?}
         */
        ChgdrTreeBuilder.prototype.rowToNode = /**
         * @private
         * @param {?} row
         * @return {?}
         */
            function (row) {
                var _this = this;
                /** @type {?} */
                var node = new ChgdrTreeNode();
                node.id = Guid.newGuid();
                node.name = this.buildTreeNodeName(row);
                node.data = {
                    id: node.id,
                    name: node.name
                };
                node.chgdrRow = row;
                node.chgdrData = this.chgdrData;
                node.children = [];
                row.entityNodes.forEach(( /**
                 * @param {?} entityNode
                 * @return {?}
                 */function (entityNode) {
                    entityNode.rows.forEach(( /**
                     * @param {?} childRow
                     * @return {?}
                     */function (childRow) {
                        /** @type {?} */
                        var childNode = _this.rowToNode(childRow);
                        childNode.parent = node;
                        node.children.push(childNode);
                    }));
                }));
                return node;
            };
        /**
         * @private
         * @param {?} row
         * @return {?}
         */
        ChgdrTreeBuilder.prototype.buildTreeNodeName = /**
         * @private
         * @param {?} row
         * @return {?}
         */
            function (row) {
                /** @type {?} */
                var dataCodeName;
                if (row.dataCodes && row.dataCodes.length > 0) {
                    dataCodeName = row.dataCodes[0].fieldValue;
                }
                if (dataCodeName) {
                    return row.entityName + "-" + dataCodeName + "【" + row.operateType.name + "】";
                }
                else {
                    return row.entityName + "【" + row.operateType.name + "】";
                }
            };
        return ChgdrTreeBuilder;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/service/chgdr-metadata.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChgdrMetadataService = /** @class */ (function () {
        function ChgdrMetadataService(http, sessionService, metadataService, injector) {
            this.http = http;
            this.sessionService = sessionService;
            this.metadataService = metadataService;
            this.injector = injector;
            this.beMetadataCache = new Map();
        }
        /** 获取BE元数据 */
        /**
         * 获取BE元数据
         * @param {?} beId
         * @return {?}
         */
        ChgdrMetadataService.prototype.getGSPBusinessEntity = /**
         * 获取BE元数据
         * @param {?} beId
         * @return {?}
         */
            function (beId) {
                var _this = this;
                if (this.beMetadataCache.has(beId)) {
                    /** @type {?} */
                    var be = this.beMetadataCache.get(beId);
                    return rxjs.of(be);
                }
                else {
                    return this.metadataService.GetMetadataRT(beId).pipe(operators.map(( /**
                     * @param {?} metadata
                     * @return {?}
                     */function (metadata) {
                        /** @type {?} */
                        var be = JSON.parse(metadata.content);
                        _this.beMetadataCache.set(beId, be);
                        return be;
                    })));
                }
            };
        ChgdrMetadataService.decorators = [
            { type: i0.Injectable, args: [{ providedIn: 'root' },] }
        ];
        /** @nocollapse */
        ChgdrMetadataService.ctorParameters = function () {
            return [
                { type: i1.HttpService },
                { type: i1.SessionService },
                { type: i2.GSPMetadataRTService },
                { type: i0.Injector }
            ];
        };
        /** @nocollapse */ ChgdrMetadataService.ngInjectableDef = i0.defineInjectable({ factory: function ChgdrMetadataService_Factory() { return new ChgdrMetadataService(i0.inject(i1.HttpService), i0.inject(i1.SessionService), i0.inject(i2.GSPMetadataRTService), i0.inject(i0.INJECTOR)); }, token: ChgdrMetadataService, providedIn: "root" });
        return ChgdrMetadataService;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/service/chgdr-config.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var ServerIP = '/';
    /** @type {?} */
    var chgdrUrl = ServerIP + "api/runtime/chgdr/v1.0";
    var ChgdrConfigService = /** @class */ (function () {
        function ChgdrConfigService(http, sessionService, injector) {
            this.http = http;
            this.sessionService = sessionService;
            this.injector = injector;
        }
        /**
         * 根据业务实体ID获取所有的配置信息
         */
        /**
         * 根据业务实体ID获取所有的配置信息
         * @param {?} beId
         * @return {?}
         */
        ChgdrConfigService.prototype.getBeChgdrConfig = /**
         * 根据业务实体ID获取所有的配置信息
         * @param {?} beId
         * @return {?}
         */
            function (beId) {
                /** @type {?} */
                var queryUrl = chgdrUrl + "/config?beId=" + beId;
                return (( /** @type {?} */(( /** @type {?} */(this.http.get(queryUrl)))))).pipe(operators.tap(( /**
                 * @param {?} data
                 * @return {?}
                 */function (data) {
                    /** @type {?} */
                    var configEntities = data || [];
                    configEntities.forEach(( /**
                     * @param {?} entity
                     * @return {?}
                     */function (entity) {
                        //处理字段顺序号
                        /** @type {?} */
                        var fields = entity.fields || [];
                        //赋予老数据默认顺序号
                        fields.forEach(( /**
                         * @param {?} field
                         * @param {?} index
                         * @return {?}
                         */function (field, index) {
                            if (field.showIndex === undefined || field.showIndex === null) {
                                field.showIndex = index;
                            }
                        }));
                        //按照顺序号排序
                        entity.fields = fields.sort(( /**
                         * @param {?} a
                         * @param {?} b
                         * @return {?}
                         */function (a, b) { return a.showIndex - b.showIndex; }));
                        //根据排序结果重置顺序号
                        fields.forEach(( /**
                         * @param {?} field
                         * @param {?} index
                         * @return {?}
                         */function (field, index) {
                            field.showIndex = index;
                        }));
                    }));
                })));
            };
        ChgdrConfigService.decorators = [
            { type: i0.Injectable, args: [{ providedIn: 'root' },] }
        ];
        /** @nocollapse */
        ChgdrConfigService.ctorParameters = function () {
            return [
                { type: i1.HttpService },
                { type: i1.SessionService },
                { type: i0.Injector }
            ];
        };
        /** @nocollapse */ ChgdrConfigService.ngInjectableDef = i0.defineInjectable({ factory: function ChgdrConfigService_Factory() { return new ChgdrConfigService(i0.inject(i1.HttpService), i0.inject(i1.SessionService), i0.inject(i0.INJECTOR)); }, token: ChgdrConfigService, providedIn: "root" });
        return ChgdrConfigService;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/service/chgdr-config-handler.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChgdrConfigHandler = /** @class */ (function () {
        function ChgdrConfigHandler(be, entityConfigs) {
            this.entityCodeMap = new Map();
            this.entityFieldMap = new Map();
            this.entityFieldShowOrderMap = new Map();
            this.entityFieldParentMap = new Map();
            this.entityFieldConfigMap = new Map();
            //初始化所有BE对象信息
            this.be = be;
            if (!!be) {
                /** @type {?} */
                var mainObject = be.MainObject;
                this.updateEntityMap(mainObject, this.entityCodeMap, this.entityFieldMap);
            }
            //初始化配置信息
            this.entityConfigs = entityConfigs || [];
            this.updateEntityConfigs();
            //更新显示顺序
            this.updateShowIndex();
        }
        /** 更新显示顺序号 */
        /**
         * 更新显示顺序号
         * @private
         * @return {?}
         */
        ChgdrConfigHandler.prototype.updateShowIndex = /**
         * 更新显示顺序号
         * @private
         * @return {?}
         */
            function () {
                /** @type {?} */
                var entityFieldShowOrderMap = new Map();
                //首先用BE字段顺序赋予初始顺序，避免配置中无顺序号造成显示错乱
                this.entityFieldMap.forEach(( /**
                 * @param {?} value
                 * @param {?} key
                 * @return {?}
                 */function (value, key) {
                    entityFieldShowOrderMap.set(key, entityFieldShowOrderMap.size);
                }));
                //然后用配置顺序替代初始顺序
                this.entityFieldConfigMap.forEach(( /**
                 * @param {?} value
                 * @param {?} key
                 * @return {?}
                 */function (value, key) {
                    if (value.showIndex !== undefined && value.showIndex !== null) {
                        entityFieldShowOrderMap.set(key, value.showIndex);
                    }
                }));
                this.entityFieldShowOrderMap = entityFieldShowOrderMap;
            };
        /** 更新配置信息的相关数据 */
        /**
         * 更新配置信息的相关数据
         * @private
         * @return {?}
         */
        ChgdrConfigHandler.prototype.updateEntityConfigs = /**
         * 更新配置信息的相关数据
         * @private
         * @return {?}
         */
            function () {
                /** @type {?} */
                var fieldConfigMap = new Map();
                this.entityConfigs.forEach(( /**
                 * @param {?} entityConfig
                 * @return {?}
                 */function (entityConfig) {
                    /** @type {?} */
                    var tabCode = entityConfig.tabCode;
                    (entityConfig.fields || []).forEach(( /**
                     * @param {?} field
                     * @return {?}
                     */function (field) {
                        /** @type {?} */
                        var fieldLabel = field.fieldLabelId;
                        /** @type {?} */
                        var key = tabCode + "#" + fieldLabel;
                        fieldConfigMap.set(key, field);
                    }));
                }));
                this.entityFieldConfigMap = fieldConfigMap;
            };
        /**
         * @private
         * @param {?} entity
         * @param {?} entityCodeMap
         * @param {?} entityFieldMap
         * @return {?}
         */
        ChgdrConfigHandler.prototype.updateEntityMap = /**
         * @private
         * @param {?} entity
         * @param {?} entityCodeMap
         * @param {?} entityFieldMap
         * @return {?}
         */
            function (entity, entityCodeMap, entityFieldMap) {
                var _this = this;
                if (!entity) {
                    return;
                }
                entityCodeMap.set(entity.Code, entity);
                //实体对象的字段信息
                entity.ContainElements.forEach(( /**
                 * @param {?} ele
                 * @return {?}
                 */function (ele) {
                    _this.updateEntityFieldMap(entity, ele, null, entityFieldMap);
                }));
                //处理子节点信息
                /** @type {?} */
                var children = entity.ContainChildObjects;
                if (!children || !children.length) {
                    return;
                }
                children.forEach(( /**
                 * @param {?} childEntity
                 * @return {?}
                 */function (childEntity) {
                    _this.updateEntityMap(childEntity, entityCodeMap, entityFieldMap);
                }));
            };
        /** 更新BE节点的指定字段信息，如果该字段是关联字段，同时更新其带出字段 */
        /**
         * 更新BE节点的指定字段信息，如果该字段是关联字段，同时更新其带出字段
         * @private
         * @param {?} entity
         * @param {?} element
         * @param {?} parentElement
         * @param {?} entityFieldMap
         * @return {?}
         */
        ChgdrConfigHandler.prototype.updateEntityFieldMap = /**
         * 更新BE节点的指定字段信息，如果该字段是关联字段，同时更新其带出字段
         * @private
         * @param {?} entity
         * @param {?} element
         * @param {?} parentElement
         * @param {?} entityFieldMap
         * @return {?}
         */
            function (entity, element, parentElement, entityFieldMap) {
                var _this = this;
                /** @type {?} */
                var key = entity.Code + "#" + element.LabelID;
                entityFieldMap.set(key, element);
                if (parentElement) {
                    this.entityFieldParentMap.set(key, parentElement);
                }
                /** @type {?} */
                var childAssociations = element.ChildAssociations;
                if (!!childAssociations && childAssociations.length > 0) {
                    childAssociations.forEach(( /**
                     * @param {?} childAssociation
                     * @return {?}
                     */function (childAssociation) {
                        /** @type {?} */
                        var refElementCollection = childAssociation.RefElementCollection;
                        if (!!refElementCollection && refElementCollection.length > 0) {
                            refElementCollection.forEach(( /**
                             * @param {?} rele
                             * @return {?}
                             */function (rele) {
                                _this.updateEntityFieldMap(entity, rele, element, entityFieldMap);
                            }));
                        }
                    }));
                }
            };
        /**
         * @param {?} items
         * @param {?} entityCode
         * @param {?} fieldLabelIdFunc
         * @return {?}
         */
        ChgdrConfigHandler.prototype.sort = /**
         * @param {?} items
         * @param {?} entityCode
         * @param {?} fieldLabelIdFunc
         * @return {?}
         */
            function (items, entityCode, fieldLabelIdFunc) {
                var _this = this;
                items.sort(( /**
                 * @param {?} a
                 * @param {?} b
                 * @return {?}
                 */function (a, b) {
                    /** @type {?} */
                    var aKey = entityCode + "#" + fieldLabelIdFunc(a);
                    /** @type {?} */
                    var bKey = entityCode + "#" + fieldLabelIdFunc(b);
                    if (!_this.entityFieldShowOrderMap.has(aKey) && !_this.entityFieldShowOrderMap.has(bKey)) {
                        //如果双方均无顺序号，则按照自然排序
                        return aKey.localeCompare(bKey);
                    }
                    else if (!_this.entityFieldShowOrderMap.has(aKey)) {
                        //如果有一方无顺序号，则无顺序号的靠后
                        return -1;
                    }
                    else if (!_this.entityFieldShowOrderMap.has(bKey)) {
                        //如果有一方无顺序号，则无顺序号的靠后
                        return 1;
                    }
                    else {
                        //如果两方都有顺序号，则按照顺序号排序
                        /** @type {?} */
                        var aOrder = _this.entityFieldShowOrderMap.get(aKey);
                        /** @type {?} */
                        var bOrder = _this.entityFieldShowOrderMap.get(bKey);
                        return aOrder - bOrder;
                    }
                }));
            };
        /** 格式化主对象节点中字段的值 */
        /**
         * 格式化主对象节点中字段的值
         * @param {?} originalValue
         * @param {?} fieldLabelId
         * @return {?}
         */
        ChgdrConfigHandler.prototype.formatMainObjectFieldValue = /**
         * 格式化主对象节点中字段的值
         * @param {?} originalValue
         * @param {?} fieldLabelId
         * @return {?}
         */
            function (originalValue, fieldLabelId) {
                /** @type {?} */
                var entityCode = this.getMainObjectCode();
                return this.formatFieldValue(originalValue, entityCode, fieldLabelId);
            };
        /** 格式化字段值 */
        /**
         * 格式化字段值
         * @param {?} originalValue
         * @param {?} entityCode
         * @param {?} fieldLabelId
         * @return {?}
         */
        ChgdrConfigHandler.prototype.formatFieldValue = /**
         * 格式化字段值
         * @param {?} originalValue
         * @param {?} entityCode
         * @param {?} fieldLabelId
         * @return {?}
         */
            function (originalValue, entityCode, fieldLabelId) {
                /** @type {?} */
                var element = this.getElement(entityCode, fieldLabelId);
                if (!element) {
                    console.error("未找到业务实体【" + this.be.ID + "-" + this.be.Name + "】节点【" + entityCode + "】的字段【" + fieldLabelId + "】");
                    return originalValue;
                }
                if (!element) {
                    return originalValue;
                }
                try {
                    //处理历史遗留的数字格式
                    if ((element.MDataType == gspCmMetadata.GSPElementDataType.Decimal || element.MDataType == gspCmMetadata.GSPElementDataType.Integer)
                        && originalValue && originalValue.toLowerCase().includes("e")) {
                        //如果字段是数字类型，且是科学计数法格式
                        //转为十进制格式显示
                        /** @type {?} */
                        var num = Number(originalValue);
                        if (isNaN(num)) {
                            console.log("业务实体【" + this.be.ID + "-" + this.be.Name + "】节点【" + entityCode + "】的字段【" + fieldLabelId + "】的值不是有效的数字格式：" + originalValue);
                            return originalValue;
                        }
                        else {
                            /** @type {?} */
                            var result = this.toNonExponential(num);
                            return result;
                        }
                    }
                    //处理布尔值类型
                    if (element.MDataType == gspCmMetadata.GSPElementDataType.Boolean) {
                        //注意：new Boolean("false") == true
                        if (originalValue === null || originalValue === undefined || originalValue.length == 0) {
                            return originalValue;
                        }
                        else if (originalValue.toLowerCase() == "false") {
                            return I18nUtil.instantOrDefault("no-text", "否");
                        }
                        else if (originalValue.toLowerCase() == "true") {
                            return I18nUtil.instantOrDefault("yes-text", "是");
                        }
                        else {
                            return originalValue;
                        }
                    }
                    //处理日期格式
                    if (element.MDataType == gspCmMetadata.GSPElementDataType.Date) {
                        if (originalValue === null || originalValue === undefined || originalValue.length == 0) {
                            //不包含值，直接返回原始值
                            return originalValue;
                        }
                        else {
                            /** @type {?} */
                            var date = new Date(originalValue);
                            /** @type {?} */
                            var str = dateFns.format(date, "yyyy-MM-dd");
                            return str;
                        }
                    }
                    //处理日期时间格式
                    if (element.MDataType == gspCmMetadata.GSPElementDataType.DateTime) {
                        if (originalValue === null || originalValue === undefined || originalValue.length == 0) {
                            //不包含值，直接返回原始值
                            return originalValue;
                        }
                        else {
                            /** @type {?} */
                            var date = new Date(originalValue);
                            /** @type {?} */
                            var str = dateFns.format(date, "yyyy-MM-dd HH:mm:ss");
                            return str;
                        }
                    }
                    //处理枚举格式
                    if (element.ObjectType == gspCmMetadata.GSPElementObjectType.Enum) {
                        /** @type {?} */
                        var enumValues = element.ContainEnumValues || [];
                        /** @type {?} */
                        var enumValue = enumValues.find(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) { return item.Value == originalValue; }));
                        if (enumValue) {
                            return enumValue.Name;
                        }
                        else if (originalValue === null || originalValue === undefined || originalValue.length == 0) {
                            //如果找不到对应的枚举项，并且原始值为空，说明未赋值，不报错
                            return originalValue;
                        }
                        else {
                            //打印异常
                            console.log("业务实体【" + this.be.ID + "-" + this.be.Name + "】节点【" + entityCode + "】的字段【" + fieldLabelId + "】的值未找到匹配的枚举项：" + originalValue);
                            return originalValue;
                        }
                    }
                    return originalValue;
                }
                catch (error) {
                    console.error("业务实体【" + this.be.ID + "-" + this.be.Name + "】节点【" + entityCode + "】的字段【" + fieldLabelId + "】的值【" + originalValue + "】格式化出错：", error);
                    return originalValue;
                }
            };
        /** JavaScript中科学计数法转化为数值字符串形式。
         * 以下两种情况，JavaScript 会自动将数值转为科学计数法表示
         * (1) 小于1且小数点后面带有6个0以上的浮点数值：
         * (2) 整数位数字多于21位：
         * 首先看看整数位数字多于21位的情况，其实这个一般不会碰到，整数位数字多于21位已经超出了 JavaScript 精确整数范围。
         * 这个方法用于处理小数问题
         * from：https://www.cnblogs.com/xianfengzhike/p/9939029.html */
        /**
         * JavaScript中科学计数法转化为数值字符串形式。
         * 以下两种情况，JavaScript 会自动将数值转为科学计数法表示
         * (1) 小于1且小数点后面带有6个0以上的浮点数值：
         * (2) 整数位数字多于21位：
         * 首先看看整数位数字多于21位的情况，其实这个一般不会碰到，整数位数字多于21位已经超出了 JavaScript 精确整数范围。
         * 这个方法用于处理小数问题
         * from：https://www.cnblogs.com/xianfengzhike/p/9939029.html
         * @private
         * @param {?} num
         * @return {?}
         */
        ChgdrConfigHandler.prototype.toNonExponential = /**
         * JavaScript中科学计数法转化为数值字符串形式。
         * 以下两种情况，JavaScript 会自动将数值转为科学计数法表示
         * (1) 小于1且小数点后面带有6个0以上的浮点数值：
         * (2) 整数位数字多于21位：
         * 首先看看整数位数字多于21位的情况，其实这个一般不会碰到，整数位数字多于21位已经超出了 JavaScript 精确整数范围。
         * 这个方法用于处理小数问题
         * from：https://www.cnblogs.com/xianfengzhike/p/9939029.html
         * @private
         * @param {?} num
         * @return {?}
         */
            function (num) {
                /** @type {?} */
                var m = num.toExponential().match(/\d(?:\.(\d*))?e([+-]\d+)/);
                return num.toFixed(Math.max(0, (m[1] || '').length - m[2]));
            };
        /**
         * @return {?}
         */
        ChgdrConfigHandler.prototype.getBeName = /**
         * @return {?}
         */
            function () {
                return this.be ? this.be.Name : null;
            };
        /**
         * @return {?}
         */
        ChgdrConfigHandler.prototype.getMainObjectCode = /**
         * @return {?}
         */
            function () {
                return this.be && this.be.MainObject ? this.be.MainObject.Code : null;
            };
        /**
         * @param {?} entityCode
         * @param {?} fieldLabelId
         * @return {?}
         */
        ChgdrConfigHandler.prototype.getElement = /**
         * @param {?} entityCode
         * @param {?} fieldLabelId
         * @return {?}
         */
            function (entityCode, fieldLabelId) {
                /** @type {?} */
                var key = entityCode + "#" + fieldLabelId;
                return this.entityFieldMap.get(key);
            };
        /**
         * @param {?} entityCode
         * @param {?} fieldLabelId
         * @return {?}
         */
        ChgdrConfigHandler.prototype.getParentElement = /**
         * @param {?} entityCode
         * @param {?} fieldLabelId
         * @return {?}
         */
            function (entityCode, fieldLabelId) {
                /** @type {?} */
                var key = entityCode + "#" + fieldLabelId;
                return this.entityFieldParentMap.get(key);
            };
        /**
         * @param {?} entityCode
         * @return {?}
         */
        ChgdrConfigHandler.prototype.getEntityNameOrDefault = /**
         * @param {?} entityCode
         * @return {?}
         */
            function (entityCode) {
                /** @type {?} */
                var entity = this.entityCodeMap.get(entityCode);
                return !!entity ? entity.Name : entityCode;
            };
        /**
         * @param {?} entityCode
         * @param {?} fieldLabelId
         * @return {?}
         */
        ChgdrConfigHandler.prototype.getEntityFieldNameOrDefault = /**
         * @param {?} entityCode
         * @param {?} fieldLabelId
         * @return {?}
         */
            function (entityCode, fieldLabelId) {
                /** @type {?} */
                var key = entityCode + "#" + fieldLabelId;
                /** @type {?} */
                var fieldConfig = this.entityFieldConfigMap.get(key);
                if (fieldConfig && fieldConfig.fieldName && fieldConfig.fieldName != "") {
                    //优先返回配置的名称
                    return fieldConfig.fieldName;
                }
                //如果不存在配置名称，则返回BE字段名称
                /** @type {?} */
                var ele = this.entityFieldMap.get(key);
                return !!ele ? ele.Name : fieldLabelId;
            };
        return ChgdrConfigHandler;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/service/chgdr-inner.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChgdrInnerService = /** @class */ (function () {
        function ChgdrInnerService(http, sessionService, metadataService, chgdrMetadataService, chgdrConfigService, injector) {
            this.http = http;
            this.sessionService = sessionService;
            this.metadataService = metadataService;
            this.chgdrMetadataService = chgdrMetadataService;
            this.chgdrConfigService = chgdrConfigService;
            this.injector = injector;
            this.beMetadataCache = new Map();
        }
        /** 获取BE元数据 */
        /**
         * 获取BE元数据
         * @param {?} beId
         * @return {?}
         */
        ChgdrInnerService.prototype.getChgdrConfigHandler = /**
         * 获取BE元数据
         * @param {?} beId
         * @return {?}
         */
            function (beId) {
                var _this = this;
                if (this.beMetadataCache.has(beId)) {
                    /** @type {?} */
                    var be = this.beMetadataCache.get(beId);
                    return rxjs.of(be);
                }
                else {
                    /** @type {?} */
                    var requests = [];
                    requests.push(this.chgdrMetadataService.getGSPBusinessEntity(beId));
                    requests.push(this.chgdrConfigService.getBeChgdrConfig(beId));
                    return rxjs.forkJoin(requests).pipe(operators.map(( /**
                     * @param {?} data
                     * @return {?}
                     */function (data) {
                        /** @type {?} */
                        var be = data[0];
                        /** @type {?} */
                        var entityConfigs = data[1] || [];
                        /** @type {?} */
                        var handler = new ChgdrConfigHandler(be, entityConfigs);
                        _this.beMetadataCache.set(beId, handler);
                        return handler;
                    })));
                }
            };
        ChgdrInnerService.decorators = [
            { type: i0.Injectable, args: [{ providedIn: 'root' },] }
        ];
        /** @nocollapse */
        ChgdrInnerService.ctorParameters = function () {
            return [
                { type: i1.HttpService },
                { type: i1.SessionService },
                { type: i2.GSPMetadataRTService },
                { type: ChgdrMetadataService },
                { type: ChgdrConfigService },
                { type: i0.Injector }
            ];
        };
        /** @nocollapse */ ChgdrInnerService.ngInjectableDef = i0.defineInjectable({ factory: function ChgdrInnerService_Factory() { return new ChgdrInnerService(i0.inject(i1.HttpService), i0.inject(i1.SessionService), i0.inject(i2.GSPMetadataRTService), i0.inject(ChgdrMetadataService), i0.inject(ChgdrConfigService), i0.inject(i0.INJECTOR)); }, token: ChgdrInnerService, providedIn: "root" });
        return ChgdrInnerService;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/chgdr-data-viewer/chgdr-data-viewer.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChgdrDataViewerComponent = /** @class */ (function () {
        function ChgdrDataViewerComponent(metadataService, injector, chgdrMetadataService, chgdrConfigService, chgdrInnerService) {
            this.metadataService = metadataService;
            this.injector = injector;
            this.chgdrMetadataService = chgdrMetadataService;
            this.chgdrConfigService = chgdrConfigService;
            this.chgdrInnerService = chgdrInnerService;
            this.showDataId = false;
            this.treeConfig = {
                idField: 'id',
                columns: [
                    { field: 'name', title: 'Name', width: 200 }
                ],
                /**
                 * 叶子节点图标
                 */
                leafIcon: 'f-icon f-icon-page-title-type text-info mr-2',
                /**
                 * 节点展开图标
                 */
                expandIcon: 'f-icon f-icon-file-folder-open text-info mr-2',
                /**
                 * 节点折叠图标
                 */
                collapseIcon: 'f-icon f-icon-file-folder-close text-info mr-2',
                showLevel: -1,
                disabled: false,
                init: false,
                selectValues: []
            };
        }
        /**
         * @return {?}
         */
        ChgdrDataViewerComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        ChgdrDataViewerComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                var _this = this;
                if (!!this.data) {
                    /** @type {?} */
                    var beId = this.data.entityId;
                    this.chgdrInnerService.getChgdrConfigHandler(beId).subscribe(( /**
                     * @param {?} handler
                     * @return {?}
                     */function (handler) {
                        _this.chgdrData = new ChgdrDataBuilder().buildChgdrData(_this.data, handler);
                        _this.treeData = new ChgdrTreeBuilder().build(_this.chgdrData);
                        if (_this.treeTableComponent) {
                            //移除已选中项。
                            //避免更换数据源后，treeGrid找不到之前的选中数据，导致控制台告警
                            _this.treeTableComponent.clearSelections();
                        }
                        setTimeout(( /**
                         * @return {?}
                         */function () {
                            _this.treeTableComponent.selectNode(_this.treeData[0].id);
                        }), 0);
                    }));
                }
            };
        /**
         * @param {?} $event
         * @return {?}
         */
        ChgdrDataViewerComponent.prototype.handleTreeSelection = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                /** @type {?} */
                var node = $event.node;
                if (this.currentNode != null && this.currentNode.id == node.id) {
                    //如果选中相同节点，直接返回
                    return;
                }
                this.currentNode = node;
            };
        Object.defineProperty(ChgdrDataViewerComponent.prototype, "baseInfoSectionTitle", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instant("base-info-section-title");
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ChgdrDataViewerComponent.prototype, "changeDetailSectionTitle", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instant("change-detail-section-title");
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ChgdrDataViewerComponent.prototype, "emptyDataText", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instant("empty-data-text");
            },
            enumerable: true,
            configurable: true
        });
        ChgdrDataViewerComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'chgdr-data-viewer',
                        template: "<farris-splitter *ngIf=\"!!chgdrData\" class=\"row flex-fill farris-split-section\" style=\"margin: 0px;\">\r\n  <farris-splitter-pane [ngResizable]=true [rzMinWidth]=\"250\" [rzHandles]=\"'e'\" style=\"padding:0.3rem 0 0.3rem 0.3rem;\"\r\n    class=\"d-flex flex-column farris-lsection f-col-w4\">\r\n    <farris-treetable #treeTable class=\"f-utils-fill\" style=\"overflow-y: auto;\" [columns]=\"treeConfig.columns\" [showBorder]=\"false\" [showHeader]=\"false\"\r\n      [data]=\"treeData\" [idField]=\"treeConfig.idField\" [singleSelect]=\"true\" [showCheckbox]=\"false\"\r\n      [showIcon]=\"true\" [disabled]=\"treeConfig.disabled\" [checkOnSelect]=\"true\"\r\n      (nodeSelected)=\"handleTreeSelection($event)\" [leafIcon]=\"treeConfig.leafIcon\"\r\n      [expandIcon]=\"treeConfig.expandIcon\" [collapseIcon]=\"treeConfig.collapseIcon\"\r\n      [selectValues]=\"treeConfig.selectValues\">\r\n    </farris-treetable>\r\n  </farris-splitter-pane>\r\n  <farris-splitter-pane class=\"farris-rsection farris-overflow-y-auto f-utils-fill-flex-column\" style=\"padding:0 0 0 0.3rem\">\r\n    <farris-section [mainTitle]=\"baseInfoSectionTitle\" [enableMaximize]=\"false\" [enableAccordion]=\"'default'\">\r\n      <lib-chgdr-row-baseinfo-form [chgdrTreeNode]=\"currentNode\" [showDataId]=\"showDataId\"></lib-chgdr-row-baseinfo-form>\r\n    </farris-section>\r\n    <farris-section [mainTitle]=\"changeDetailSectionTitle\" [enableMaximize]=\"false\" [fill]=\"true\" class=\"pt-0\">\r\n      <lib-chgdr-row-grid *ngIf=\"!!currentNode && currentNode.chgdrRow.changes.length>0\" [chgdrTreeNode]=\"currentNode\" class=\"f-utils-fill\"></lib-chgdr-row-grid>\r\n      <div class=\"f-datagrid-norecords\" *ngIf=\"!currentNode || currentNode.chgdrRow.changes.length==0\" style=\"height: 250px;\">\r\n        <div class=\"f-datagrid-norecords-content\" style=\"width: 100%;\">\r\n            {{emptyDataText}}\r\n        </div>\r\n      </div>\r\n    </farris-section>\r\n  </farris-splitter-pane>\r\n</farris-splitter>",
                        styles: [":host{display:flex;height:100%}"]
                    }] }
        ];
        /** @nocollapse */
        ChgdrDataViewerComponent.ctorParameters = function () {
            return [
                { type: i2.GSPMetadataRTService },
                { type: i0.Injector },
                { type: ChgdrMetadataService },
                { type: ChgdrConfigService },
                { type: ChgdrInnerService }
            ];
        };
        ChgdrDataViewerComponent.propDecorators = {
            data: [{ type: i0.Input }],
            showDataId: [{ type: i0.Input }],
            treeTableComponent: [{ type: i0.ViewChild, args: ["treeTable",] }]
        };
        return ChgdrDataViewerComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/model/operate-type.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var OperateType = /** @class */ (function () {
        function OperateType(ordinal, code, defaultName) {
            this.ordinal = ordinal;
            this.code = code;
            this.defaultName = defaultName;
        }
        Object.defineProperty(OperateType.prototype, "name", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instantOrDefault("operate-type-name-" + this.code, this.defaultName);
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        OperateType.values = /**
         * @return {?}
         */
            function () {
                return [OperateType.ADD, OperateType.MODIFY, OperateType.DELETE];
            };
        /**
         * @return {?}
         */
        OperateType.prototype.toJson = /**
         * @return {?}
         */
            function () {
                return this.code;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        OperateType.parse = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                if (obj === undefined || obj === null) {
                    return null;
                }
                if (obj instanceof OperateType) {
                    return obj;
                }
                if (typeof obj === "string") {
                    return OperateType.values().find(( /**
                     * @param {?} ot
                     * @return {?}
                     */function (ot) { return ot.code == obj; }));
                }
                if (typeof obj === "number") {
                    return OperateType.values().find(( /**
                     * @param {?} ot
                     * @return {?}
                     */function (ot) { return ot.ordinal == obj; }));
                }
                if (!!obj.code) {
                    return OperateType.values().find(( /**
                     * @param {?} ot
                     * @return {?}
                     */function (ot) { return ot.code == obj.code; }));
                }
                return null;
            };
        /**
         * 新增
         */
        OperateType.ADD = new OperateType(0, "ADD", "新增");
        /**
         * 修改
         */
        OperateType.MODIFY = new OperateType(1, "MODIFY", "修改");
        /**
         * 删除
         */
        OperateType.DELETE = new OperateType(2, "DELETE", "删除");
        return OperateType;
    }());

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m)
            return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length)
                    o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/util/map.util.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MapUtil = /** @class */ (function () {
        function MapUtil() {
        }
        /** 将Map转为Object，便于序列化。（因为JSON.stringfy(Map)返回的值总为{}） */
        /**
         * 将Map转为Object，便于序列化。（因为JSON.stringfy(Map)返回的值总为{}）
         * @param {?} map
         * @return {?}
         */
        MapUtil.convertMapToObject = /**
         * 将Map转为Object，便于序列化。（因为JSON.stringfy(Map)返回的值总为{}）
         * @param {?} map
         * @return {?}
         */
            function (map) {
                if (!map) {
                    return null;
                }
                /** @type {?} */
                var obj = Object.create(null);
                map.forEach(( /**
                 * @param {?} value
                 * @param {?} key
                 * @return {?}
                 */function (value, key) {
                    obj[key] = value;
                }));
                return obj;
            };
        /** 将Object转为Map，因为JSON.parse无法识别类型，反序列化后只能是Object类型 */
        /**
         * 将Object转为Map，因为JSON.parse无法识别类型，反序列化后只能是Object类型
         * @param {?} obj
         * @return {?}
         */
        MapUtil.convertObjectToMap = /**
         * 将Object转为Map，因为JSON.parse无法识别类型，反序列化后只能是Object类型
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                var e_1, _a;
                if (obj === undefined || obj === null) {
                    return null;
                }
                /** @type {?} */
                var strMap = new Map();
                try {
                    for (var _b = __values(Object.keys(obj)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var k = _c.value;
                        strMap.set(k, obj[k]);
                    }
                }
                catch (e_1_1) {
                    e_1 = { error: e_1_1 };
                }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return))
                            _a.call(_b);
                    }
                    finally {
                        if (e_1)
                            throw e_1.error;
                    }
                }
                return strMap;
            };
        return MapUtil;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/service/chgdr.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var ServerIP$1 = '/';
    /** @type {?} */
    var chgdrUrl$1 = ServerIP$1 + "api/runtime/chgdr/v1.0";
    var ChgdrService = /** @class */ (function () {
        function ChgdrService(http, sessionService, injector) {
            this.http = http;
            this.sessionService = sessionService;
            this.injector = injector;
            this.rootDataCodeFieldsMap = new Map();
        }
        /**
         * @param {?} currentQueryParam
         * @return {?}
         */
        ChgdrService.prototype.queryChangeDataHeader = /**
         * @param {?} currentQueryParam
         * @return {?}
         */
            function (currentQueryParam) {
                var _this = this;
                /** @type {?} */
                var obj = Object.assign({}, currentQueryParam);
                if (currentQueryParam.dataCode) {
                    obj.dataCode = MapUtil.convertMapToObject(currentQueryParam.dataCode);
                }
                /** @type {?} */
                var json = JSON.stringify(obj);
                /** @type {?} */
                var queryParam = encodeURIComponent(json);
                /** @type {?} */
                var url = chgdrUrl$1 + "?queryParam=" + queryParam;
                return this.http.get(url).pipe(operators.map(( /**
                 * @param {?} data
                 * @return {?}
                 */function (data) {
                    if (!data) {
                        return data;
                    }
                    /** @type {?} */
                    var queryResult = ( /** @type {?} */(( /** @type {?} */(data))));
                    queryResult && queryResult.headers && queryResult.headers.forEach(( /**
                     * @param {?} header
                     * @return {?}
                     */function (header) {
                        //TODO 待删，兼容老数据格式
                        if (typeof header.dataCode == "string") {
                            header.dataCode = JSON.parse(header.dataCode);
                        }
                        header.dataCode = MapUtil.convertObjectToMap(header.dataCode);
                        header.changeTime = _this.toDate(header.changeTime);
                        header.operateType = OperateType.parse(header.operateType);
                    }));
                    return queryResult;
                })));
            };
        /**
         * @param {?} beId
         * @return {?}
         */
        ChgdrService.prototype.getRootEntityDataCodeFields = /**
         * @param {?} beId
         * @return {?}
         */
            function (beId) {
                var _this = this;
                if (!beId) {
                    return rxjs.of([]);
                }
                if (this.rootDataCodeFieldsMap.has(beId)) {
                    //直接从缓存中获取
                    return rxjs.of(this.rootDataCodeFieldsMap.get(beId));
                }
                else {
                    //获取结果并存入缓存
                    /** @type {?} */
                    var url = chgdrUrl$1 + "/rootDataCodeFields?beId=" + beId;
                    return (( /** @type {?} */(( /** @type {?} */((this.http.get(url)))))))
                        .pipe(operators.tap(( /**
                 * @param {?} data
                 * @return {?}
                 */function (data) { return _this.rootDataCodeFieldsMap.set(beId, data); })));
                }
            };
        /** 将字符串或数字转为Date */
        /**
         * 将字符串或数字转为Date
         * @private
         * @param {?} date
         * @return {?}
         */
        ChgdrService.prototype.toDate = /**
         * 将字符串或数字转为Date
         * @private
         * @param {?} date
         * @return {?}
         */
            function (date) {
                if (typeof date == "string") {
                    return new Date(date);
                }
                else if (typeof date == "number") {
                    return new Date(date);
                }
                else {
                    return date;
                }
            };
        /**
         * @param {?} idOrParam
         * @param {?=} changeTime
         * @return {?}
         */
        ChgdrService.prototype.getChangeData = /**
         * @param {?} idOrParam
         * @param {?=} changeTime
         * @return {?}
         */
            function (idOrParam, changeTime) {
                if (idOrParam == null) {
                    throw new Error("id can not be null");
                }
                if (idOrParam instanceof String || typeof idOrParam == "string") {
                    return this.getChangeDataById(( /** @type {?} */(idOrParam)), changeTime);
                }
                else {
                    return this.getChangeDataByParam(idOrParam);
                }
            };
        /**
         * @private
         * @param {?} id
         * @param {?} changeTime
         * @return {?}
         */
        ChgdrService.prototype.getChangeDataById = /**
         * @private
         * @param {?} id
         * @param {?} changeTime
         * @return {?}
         */
            function (id, changeTime) {
                var _this = this;
                /** @type {?} */
                var json = JSON.stringify(changeTime);
                /** @type {?} */
                var queryParam = encodeURIComponent(json);
                /** @type {?} */
                var url = chgdrUrl$1 + "/" + id + "?changeTime=" + queryParam;
                return this.http.get(url).pipe(operators.map(( /**
                 * @param {?} data
                 * @return {?}
                 */function (data) {
                    return _this.transChangeDataFromResponse(data);
                })));
            };
        /**
         * @private
         * @param {?} param
         * @return {?}
         */
        ChgdrService.prototype.getChangeDataByParam = /**
         * @private
         * @param {?} param
         * @return {?}
         */
            function (param) {
                var _this = this;
                if (param == null) {
                    throw new Error("ChgdrItemQueryParam can not be null");
                }
                /** @type {?} */
                var requestParam = {};
                Object.assign(requestParam, param);
                //将Map转为Object，否则无法正常序列化
                requestParam.dataIds = MapUtil.convertMapToObject(param.dataIds);
                /** @type {?} */
                var url = "" + chgdrUrl$1;
                return this.http.post(url, requestParam).pipe(operators.map(( /**
                 * @param {?} data
                 * @return {?}
                 */function (data) {
                    return _this.transChangeDataFromResponse(data);
                })));
            };
        /**
         * @private
         * @param {?} data
         * @return {?}
         */
        ChgdrService.prototype.transChangeDataFromResponse = /**
         * @private
         * @param {?} data
         * @return {?}
         */
            function (data) {
                if (!data) {
                    return data;
                }
                /** @type {?} */
                var sh = ( /** @type {?} */(( /** @type {?} */(data))));
                //TODO 待删，兼容老数据格式
                if (typeof sh.dataCode == "string") {
                    sh.dataCode = JSON.parse(sh.dataCode);
                }
                sh.dataCode = MapUtil.convertObjectToMap(sh.dataCode);
                sh.changeTime = this.toDate(sh.changeTime);
                sh.operateType = OperateType.parse(sh.operateType);
                sh.rows = this.transChangeDataRowsFromResponse(sh.rows || []);
                return sh;
            };
        /**
         * @private
         * @param {?} rows
         * @return {?}
         */
        ChgdrService.prototype.transChangeDataRowsFromResponse = /**
         * @private
         * @param {?} rows
         * @return {?}
         */
            function (rows) {
                if (!rows || rows.length == 0) {
                    return rows;
                }
                rows.forEach(( /**
                 * @param {?} row
                 * @return {?}
                 */function (row) {
                    if (!row) {
                        return;
                    }
                    row.operateType = OperateType.parse(row.operateType);
                    //TODO 待删，兼容老数据格式
                    if (typeof row.dataCode == "string") {
                        row.dataCode = JSON.parse(row.dataCode);
                    }
                    row.dataCode = MapUtil.convertObjectToMap(row.dataCode);
                    row.oldContent = ( /** @type {?} */(MapUtil.convertObjectToMap(row.oldContent))) || new Map();
                    row.newContent = ( /** @type {?} */(MapUtil.convertObjectToMap(row.newContent))) || new Map();
                }));
                return rows;
            };
        ChgdrService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */
        ChgdrService.ctorParameters = function () {
            return [
                { type: i1.HttpService },
                { type: i1.SessionService },
                { type: i0.Injector }
            ];
        };
        /** @nocollapse */ ChgdrService.ngInjectableDef = i0.defineInjectable({ factory: function ChgdrService_Factory() { return new ChgdrService(i0.inject(i1.HttpService), i0.inject(i1.SessionService), i0.inject(i0.INJECTOR)); }, token: ChgdrService, providedIn: "root" });
        return ChgdrService;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/model/change-data-query-param.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChangeDataQueryParam = /** @class */ (function () {
        function ChangeDataQueryParam() {
        }
        return ChangeDataQueryParam;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/model/change-data-query-result.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChangeDataQueryResult = /** @class */ (function () {
        function ChangeDataQueryResult() {
        }
        return ChangeDataQueryResult;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/util/error.util.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ErrorUtil = /** @class */ (function () {
        function ErrorUtil() {
        }
        /**
         * @param {?} e
         * @return {?}
         */
        ErrorUtil.getErrorMessage = /**
         * @param {?} e
         * @return {?}
         */
            function (e) {
                if (!e) {
                    return "";
                }
                else if (!e.error) {
                    return "";
                }
                else if (!e.error.Message) {
                    return "";
                }
                else {
                    return e.error.Message;
                }
            };
        return ErrorUtil;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/util/time.util.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var TimeUtil = /** @class */ (function () {
        function TimeUtil() {
        }
        /**
         * @param {?} date
         * @return {?}
         */
        TimeUtil.getStartTimeInDay = /**
         * @param {?} date
         * @return {?}
         */
            function (date) {
                /** @type {?} */
                var d = new Date(date);
                d.setHours(0);
                d.setMinutes(0);
                d.setSeconds(0);
                d.setMilliseconds(0);
                return d;
            };
        /**
         * @param {?} date
         * @return {?}
         */
        TimeUtil.getEndTimeInDay = /**
         * @param {?} date
         * @return {?}
         */
            function (date) {
                /** @type {?} */
                var d = new Date(date);
                d.setHours(23);
                d.setMinutes(59);
                d.setSeconds(59);
                d.setMilliseconds(999);
                return d;
            };
        return TimeUtil;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/chgdr-list/chgdr-list.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChgdrListComponent = /** @class */ (function () {
        function ChgdrListComponent(chgdrService, messageService, notifyService, injector) {
            this.chgdrService = chgdrService;
            this.messageService = messageService;
            this.notifyService = notifyService;
            this.injector = injector;
            this.baseCls = true;
        }
        /**
         * @return {?}
         */
        ChgdrListComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var chgdrs = new ChangeDataQueryResult();
                chgdrs.total = 0;
                chgdrs.pageSize = 20;
                chgdrs.pageIndex = 1;
                chgdrs.headers = [];
                this.chgdrs = chgdrs;
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        ChgdrListComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                if (!this.changeTimeStart && !this.changeTimeEnd) {
                    //未录入变更时间范围，则查询最近一周的
                    this.changeTimeEnd = TimeUtil.getEndTimeInDay(new Date());
                    this.changeTimeStart = TimeUtil.getStartTimeInDay(new Date(new Date().getTime() - 7 * 24 * 60 * 60 * 1000));
                }
                else if (!this.changeTimeEnd) {
                    //如果未设置截至时间，则与开始时间相同
                    this.changeTimeEnd = TimeUtil.getEndTimeInDay(this.changeTimeStart);
                }
                else if (!this.changeTimeStart) {
                    //如果未设置开始时间，则与截至时间相同
                    this.changeTimeStart = TimeUtil.getEndTimeInDay(this.changeTimeEnd);
                }
                if (!!this.entityId) {
                    /** @type {?} */
                    var param = this.buildQueryParam();
                    this.doQuery(param);
                }
            };
        /**
         * @return {?}
         */
        ChgdrListComponent.prototype.query = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var param = this.buildQueryParam();
                this.doQuery(param);
            };
        /**
         * @private
         * @return {?}
         */
        ChgdrListComponent.prototype.buildQueryParam = /**
         * @private
         * @return {?}
         */
            function () {
                /** @type {?} */
                var param = new ChangeDataQueryParam();
                param.pageIndex = 1;
                param.pageSize = 20;
                param.entityId = this.entityId;
                param.dataId = this.dataId;
                param.changeTimeStart = this.changeTimeStart;
                param.changeTimeEnd = this.changeTimeEnd;
                return param;
            };
        /**
         * @private
         * @param {?} param
         * @return {?}
         */
        ChgdrListComponent.prototype.doQuery = /**
         * @private
         * @param {?} param
         * @return {?}
         */
            function (param) {
                var _this = this;
                this.chgdrService.queryChangeDataHeader(param).subscribe(( /**
                 * @param {?} data
                 * @return {?}
                 */function (data) {
                    _this.currentQueryParam = param;
                    _this.chgdrs = data;
                }), ( /**
                 * @param {?} error
                 * @return {?}
                 */function (error) {
                    console.error("查询业务变更日志出错", error);
                    _this.messageService.error(I18nUtil.instant("query-change-data-record-error", { errorMessage: ErrorUtil.getErrorMessage(error) }));
                }));
            };
        /**
         * @param {?} $event
         * @return {?}
         */
        ChgdrListComponent.prototype.onPageChanged = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                if (!$event) {
                    return;
                }
                if ($event.pageIndex) {
                    this.currentQueryParam.pageIndex = $event.pageIndex;
                }
                if ($event.pageSize) {
                    this.currentQueryParam.pageSize = $event.pageSize;
                }
                this.doQuery(this.currentQueryParam);
            };
        /**
         * @param {?} $event
         * @return {?}
         */
        ChgdrListComponent.prototype.onPageSizeChanged = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                this.currentQueryParam.pageSize = $event;
                this.doQuery(this.currentQueryParam);
            };
        Object.defineProperty(ChgdrListComponent.prototype, "startTimeLabel", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instant("start-time-label");
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ChgdrListComponent.prototype, "endTimeLabel", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instant("end-time-label");
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ChgdrListComponent.prototype, "queryBtn", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instant("query-btn");
            },
            enumerable: true,
            configurable: true
        });
        ChgdrListComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'chgdr-list',
                        template: "<div class=\"farris-form farris-form-controls-inline\" style=\"background: #fff;\">\n  <form class=\"f-section-formgroup\" style=\"padding-top: 10px;\">\n    <div class=\"f-section-formgroup-inputs\">\n      <div class=\"col-12 row\">\n        <div class=\"farris-group-wrap col-5\">\n          <div class=\"form-group farris-form-group\">\n            <label class=\"col-form-label\">\n              <span class=\"farris-label-info text-danger\">*</span><span class=\"farris-label-text\">{{startTimeLabel}}</span>\n            </label>\n            <div class=\"farris-input-wrap\">\n              <farris-datepicker [(ngModel)]=\"changeTimeStart\" name=\"changeTimeStart\"></farris-datepicker>\n            </div>\n          </div>\n        </div>\n        <div class=\"farris-group-wrap col-5\">\n          <div class=\"form-group farris-form-group\">\n            <label class=\"col-form-label\">\n              <span class=\"farris-label-info text-danger\">*</span><span class=\"farris-label-text\">{{endTimeLabel}}</span>\n            </label>\n            <div class=\"farris-input-wrap\">\n              <farris-datepicker [(ngModel)]=\"changeTimeEnd\" name=\"changeTimeEnd\"></farris-datepicker>\n            </div>\n          </div>\n        </div>\n        <div class=\"farris-group-wrap col-2\">\n          <div class=\"form-group farris-form-group\">\n            <label class=\"col-form-label\">\n              <button class=\"btn btn-primary mr-2\" (click)=\"query()\">{{queryBtn}}</button>\n            </label>\n            <div class=\"farris-input-wrap\">\n            </div>\n          </div>\n        </div>\n      </div>\n    </div>\n  </form>\n</div>\n\n<lib-chgdr-header-datagrid\n  [chgdrs]=\"chgdrs\" \n  (pageChanged)=\"onPageChanged($event)\" \n  (pageSizeChanged)=\"onPageSizeChanged($event)\"\n  [showDataId]=\"false\"\n  [beforeShowChangeDetail]=\"beforeShowChangeDetail\"\n></lib-chgdr-header-datagrid>\n\n<!-- <div class=\"f-utils-fill\">\n  <farris-datagrid #chgdrGrid idField=\"id\" [columns]=\"columns\" [data]=\"chgdrs.headers\"\n  showLineNumber=\"true\" [fit]=\"true\" [fitColumns]=\"true\" \n  [pagination]=\"true\" [pageSize]=\"20\" [showPageList]=\"false\" [pageIndex]=\"chgdrs.pageIndex\" [total]=\"chgdrs.total\"\n  (pageChanged)=\"onPageChanged($event)\" (pageSizeChanged)=\"onPageSizeChanged($event)\"\n  >\n  </farris-datagrid>\n  <ng-template #opCell let-ctx>\n    <a href=\"javascript: void(0);\" (click)=\"showChangeDetail(ctx)\">\u53D8\u66F4\u8BE6\u60C5</a>\n  </ng-template>\n</div> -->\n\n<!-- \u67E5\u770B\u53D8\u66F4\u65E5\u5FD7\u7684\u5F39\u7A97 -->\n<!-- <farris-dialog #chgdrInfoDialog\n  title=\"\u53D8\u66F4\u65E5\u5FD7\u8BE6\u60C5\"\n  [width]=\"1100\"\n  [height]=\"600\"\n  [showMaxButton]=\"true\"\n  [resizable]=\"true\"\n  [showButtons]=\"false\"\n>\n  <chgdr-data-viewer [data]=\"currentChgdrData\"></chgdr-data-viewer>\n</farris-dialog> -->",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        ChgdrListComponent.ctorParameters = function () {
            return [
                { type: ChgdrService },
                { type: i2$1.MessagerService },
                { type: uiNotify.NotifyService },
                { type: i0.Injector }
            ];
        };
        ChgdrListComponent.propDecorators = {
            baseCls: [{ type: i0.HostBinding, args: ['class.f-utils-fill-flex-column',] }],
            entityId: [{ type: i0.Input }],
            dataId: [{ type: i0.Input }],
            changeTimeStart: [{ type: i0.Input }],
            changeTimeEnd: [{ type: i0.Input }],
            beforeShowChangeDetail: [{ type: i0.Input }]
        };
        return ChgdrListComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/chgdr-row-baseinfo-form/chgdr-row-baseinfo-form.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChgdrRowBaseinfoFormComponent = /** @class */ (function () {
        function ChgdrRowBaseinfoFormComponent(injector) {
            this.injector = injector;
            this.showDataId = false;
        }
        /**
         * @return {?}
         */
        ChgdrRowBaseinfoFormComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
            };
        Object.defineProperty(ChgdrRowBaseinfoFormComponent.prototype, "technologyPrimaryKeyLabel", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instant("technology-primary-key-label");
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ChgdrRowBaseinfoFormComponent.prototype, "changeUserLabel", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instant("change-user-label");
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ChgdrRowBaseinfoFormComponent.prototype, "changeTimeLabel", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instant("change-time-label");
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ChgdrRowBaseinfoFormComponent.prototype, "changeOperateType", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instant("change-operate-type");
            },
            enumerable: true,
            configurable: true
        });
        ChgdrRowBaseinfoFormComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'lib-chgdr-row-baseinfo-form',
                        template: "<div *ngIf=\"!!chgdrTreeNode\" class=\"farris-form farris-form-controls-inline\">\r\n  <form class=\"f-section-formgroup\">\r\n    <div class=\"f-section-formgroup-inputs\">\r\n      <div class=\"col-12 row\">\r\n\r\n        <div class=\"farris-group-wrap col-6\" *ngIf=\"showDataId\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\">\r\n              <span class=\"farris-label-text\">{{technologyPrimaryKeyLabel}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input type=\"text\" class=\"form-control\" [ngModel]=\"chgdrTreeNode.chgdrRow.dataId\" name=\"dataId\"\r\n                [readonly]=\"true\">\r\n              <div class=\"farris-feedback valid-feedback\"></div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n\r\n        <div class=\"farris-group-wrap col-6\" *ngFor=\"let field of chgdrTreeNode.chgdrRow.dataCodes\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\">\r\n              <span class=\"farris-label-text\">{{field.fieldName}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input type=\"text\" class=\"form-control\" [ngModel]=\"field.fieldValue\" name=\"dataCode-{{field.fieldName}}\"\r\n                [readonly]=\"true\">\r\n              <div class=\"farris-feedback valid-feedback\"></div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n\r\n        <div *ngIf=\"chgdrTreeNode.isRoot\" class=\"farris-group-wrap col-6\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\">\r\n              <span class=\"farris-label-text\">{{changeUserLabel}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input type=\"text\" class=\"form-control\" [ngModel]=\"chgdrTreeNode.chgdrData.userName\" name=\"userName\"\r\n                [readonly]=\"true\">\r\n              <div class=\"farris-feedback valid-feedback\"></div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n\r\n        <div *ngIf=\"chgdrTreeNode.isRoot\" class=\"farris-group-wrap col-6\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\">\r\n              <span class=\"farris-label-text\">{{changeTimeLabel}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input type=\"text\" class=\"form-control\" value=\"{{chgdrTreeNode.chgdrData.changeTime | date:'yyyy-MM-dd HH:mm:ss'}}\" name=\"changeTime\"\r\n                [readonly]=\"true\">\r\n              <div class=\"farris-feedback valid-feedback\"></div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n\r\n        <div class=\"farris-group-wrap col-6\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\">\r\n              <span class=\"farris-label-text\">{{changeOperateType}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input type=\"text\" class=\"form-control\" [ngModel]=\"chgdrTreeNode.chgdrRow.operateType.name\" name=\"operateType\"\r\n              [readonly]=\"true\">\r\n              <div class=\"farris-feedback valid-feedback\"></div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n        \r\n      </div>\r\n    </div>\r\n  </form>\r\n</div>",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        ChgdrRowBaseinfoFormComponent.ctorParameters = function () {
            return [
                { type: i0.Injector }
            ];
        };
        ChgdrRowBaseinfoFormComponent.propDecorators = {
            chgdrTreeNode: [{ type: i0.Input }],
            showDataId: [{ type: i0.Input }]
        };
        return ChgdrRowBaseinfoFormComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/chgdr-row-grid/chgdr-row-grid.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChgdrRowGridComponent = /** @class */ (function () {
        function ChgdrRowGridComponent(injector) {
            this.injector = injector;
        }
        /**
         * @return {?}
         */
        ChgdrRowGridComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.columns = [
                    { field: 'fieldName', width: 130, title: this.fieldNameColumnTitle },
                    { field: 'oldValue', width: 130, title: this.oldValueColumnTitle, showTips: true },
                    { field: 'newValue', width: 130, title: this.newValueColumnTitle, showTips: true }
                ];
                /** @type {?} */
                var oldValueField = this.columns[1];
                /** @type {?} */
                var newValueField = this.columns[2];
                if (!!this.chgdrTreeNode) {
                    //根据初始的变更类型更改显示列的信息
                    //此处不能直接调用DataGrid组件的方法更改，因为此处尚未创建DataGrid
                    /** @type {?} */
                    var type = this.chgdrTreeNode.chgdrRow.operateType;
                    if (type == OperateType.ADD) {
                        oldValueField.visible = false;
                        newValueField.title = this.valueColumnTitle;
                    }
                    else if (type == OperateType.DELETE) {
                        newValueField.visible = false;
                        oldValueField.title = this.valueColumnTitle;
                    }
                }
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        ChgdrRowGridComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                //原来变更行类型
                /** @type {?} */
                var oldType;
                if (!changes.chgdrTreeNode) {
                    oldType = null;
                }
                else if (!changes.chgdrTreeNode.previousValue) {
                    oldType = null;
                }
                else {
                    oldType = changes.chgdrTreeNode.previousValue.chgdrRow.operateType;
                }
                //现在变更行类型
                /** @type {?} */
                var newType;
                if (!changes.chgdrTreeNode) {
                    newType = null;
                }
                else if (!changes.chgdrTreeNode.currentValue) {
                    newType = null;
                }
                else {
                    newType = changes.chgdrTreeNode.currentValue.chgdrRow.operateType;
                }
                if (oldType == newType || !this.dataGrid || !this.dataGrid.columns) ;
                else if (newType == OperateType.ADD) {
                    this.dataGrid.showColumn("newValue");
                    this.dataGrid.hideColumn("oldValue");
                    this.dataGrid.setColumnTitle("newValue", this.valueColumnTitle);
                }
                else if (newType == OperateType.DELETE) {
                    this.dataGrid.showColumn("oldValue");
                    this.dataGrid.hideColumn("newValue");
                    this.dataGrid.setColumnTitle("oldValue", this.valueColumnTitle);
                }
                else {
                    this.dataGrid.showColumn("oldValue");
                    this.dataGrid.showColumn("newValue");
                    this.dataGrid.setColumnTitle("oldValue", this.oldValueColumnTitle);
                    this.dataGrid.setColumnTitle("newValue", this.newValueColumnTitle);
                }
                if (changes.chgdrTreeNode && changes.chgdrTreeNode.currentValue) {
                    this.data = changes.chgdrTreeNode.currentValue.chgdrRow.changes;
                }
            };
        Object.defineProperty(ChgdrRowGridComponent.prototype, "fieldNameColumnTitle", {
            get: /**
             * @private
             * @return {?}
             */ function () {
                return I18nUtil.instant("field-name-column-title");
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ChgdrRowGridComponent.prototype, "valueColumnTitle", {
            get: /**
             * @private
             * @return {?}
             */ function () {
                return I18nUtil.instant("value-column-title");
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ChgdrRowGridComponent.prototype, "oldValueColumnTitle", {
            get: /**
             * @private
             * @return {?}
             */ function () {
                return I18nUtil.instant("old-value-column-title");
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ChgdrRowGridComponent.prototype, "newValueColumnTitle", {
            get: /**
             * @private
             * @return {?}
             */ function () {
                return I18nUtil.instant("new-value-column-title");
            },
            enumerable: true,
            configurable: true
        });
        ChgdrRowGridComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'lib-chgdr-row-grid',
                        template: "<farris-datagrid \r\n   #chgdrRowGrid \r\n   idField=\"fieldLabel\" \r\n   [columns]=\"columns\" \r\n   [data]=\"data\" \r\n   [fit]=\"true\"\r\n   [fitColumns]=\"true\"\r\n   [pagination]=\"false\" \r\n>",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        ChgdrRowGridComponent.ctorParameters = function () {
            return [
                { type: i0.Injector }
            ];
        };
        ChgdrRowGridComponent.propDecorators = {
            chgdrTreeNode: [{ type: i0.Input }],
            dataGrid: [{ type: i0.ViewChild, args: ["chgdrRowGrid",] }]
        };
        return ChgdrRowGridComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/chgdr-list-dialog-content/chgdr-list-dialog-content.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChgdrListDialogContentComponent = /** @class */ (function () {
        function ChgdrListDialogContentComponent(injector) {
            this.injector = injector;
        }
        /**
         * @return {?}
         */
        ChgdrListDialogContentComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
            };
        ChgdrListDialogContentComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'lib-chgdr-list-dialog-content',
                        template: "<chgdr-list [entityId]=\"entityId\" [dataId]=\"dataId\" style=\"padding: 0.75rem 0.875rem;height:100%\" [beforeShowChangeDetail]=\"beforeShowChangeDetail\"></chgdr-list>",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        ChgdrListDialogContentComponent.ctorParameters = function () {
            return [
                { type: i0.Injector }
            ];
        };
        ChgdrListDialogContentComponent.propDecorators = {
            entityId: [{ type: i0.Input }],
            dataId: [{ type: i0.Input }],
            beforeShowChangeDetail: [{ type: i0.Input }]
        };
        return ChgdrListDialogContentComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/model/chgdr-ui-config.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BeforeShowChangeDetailEvent = /** @class */ (function () {
        function BeforeShowChangeDetailEvent() {
        }
        return BeforeShowChangeDetailEvent;
    }());
    /**
     * 对于业务变更日志对话框的设置
     */
    var /**
     * 对于业务变更日志对话框的设置
     */ ChgdrListDialogConfig = /** @class */ (function () {
        function ChgdrListDialogConfig() {
        }
        return ChgdrListDialogConfig;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/service/chgdr-list-ui.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChgdrListUiService = /** @class */ (function () {
        function ChgdrListUiService(resolver, injector, modalService, messagerService) {
            this.resolver = resolver;
            this.injector = injector;
            this.modalService = modalService;
            this.messagerService = messagerService;
        }
        /**
         * @param {?} arg1
         * @param {?=} dataId
         * @return {?}
         */
        ChgdrListUiService.prototype.showDialog = /**
         * @param {?} arg1
         * @param {?=} dataId
         * @return {?}
         */
            function (arg1, dataId) {
                if (!arg1) {
                    console.error("创建业务变更日志列表组件失败，业务实体不能为空，请检查", arg1, dataId);
                    throw new Error("创建业务变更日志列表组件失败，业务实体不能为空，请检查");
                }
                if (arg1 instanceof String || typeof arg1 == "string") {
                    /** @type {?} */
                    var config = new ChgdrListDialogConfig();
                    config.entityId = arg1.toString();
                    config.dataId = dataId;
                    this.create(config);
                }
                else {
                    this.create(arg1);
                }
            };
        /**
         * @private
         * @param {?} config
         * @return {?}
         */
        ChgdrListUiService.prototype.create = /**
         * @private
         * @param {?} config
         * @return {?}
         */
            function (config) {
                if (config.entityId == null) {
                    console.error("创建业务变更日志列表组件失败，业务实体不能为空，请检查", config);
                    throw new Error("创建业务变更日志列表组件失败，业务实体不能为空，请检查");
                }
                /** @type {?} */
                var fac = this.resolver.resolveComponentFactory(ChgdrListDialogContentComponent);
                /** @type {?} */
                var cmpRef = fac.create(this.injector);
                /** @type {?} */
                var modalConfig = {
                    title: I18nUtil.instant("chgdr-list-dialog-title"),
                    width: 1000,
                    height: 550,
                    resizable: true,
                    showButtons: false,
                    showMaxButton: true,
                    beforeOpen: ( /**
                     * @return {?}
                     */function () {
                        return rxjs.of(true);
                    })
                };
                cmpRef.instance.entityId = config.entityId;
                cmpRef.instance.dataId = config.dataId;
                cmpRef.instance.beforeShowChangeDetail = config.beforeShowChangeDetail;
                /** @type {?} */
                var dialog = this.modalService.show(cmpRef, modalConfig);
            };
        ChgdrListUiService.decorators = [
            { type: i0.Injectable, args: [{ providedIn: 'root' },] }
        ];
        /** @nocollapse */
        ChgdrListUiService.ctorParameters = function () {
            return [
                { type: i0.ComponentFactoryResolver },
                { type: i0.Injector },
                { type: i1$1.BsModalService },
                { type: i2$1.MessagerService }
            ];
        };
        /** @nocollapse */ ChgdrListUiService.ngInjectableDef = i0.defineInjectable({ factory: function ChgdrListUiService_Factory() { return new ChgdrListUiService(i0.inject(i0.ComponentFactoryResolver), i0.inject(i0.INJECTOR), i0.inject(i1$1.BsModalService), i0.inject(i2$1.MessagerService)); }, token: ChgdrListUiService, providedIn: "root" });
        return ChgdrListUiService;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/chgdr-header-datagrid/chgdr-header-datagrid.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChgdrHeaderDatagridComponent = /** @class */ (function () {
        function ChgdrHeaderDatagridComponent(chgdrService, messageService, notifyService, ref, chgdrInnerService, injector) {
            this.chgdrService = chgdrService;
            this.messageService = messageService;
            this.notifyService = notifyService;
            this.ref = ref;
            this.chgdrInnerService = chgdrInnerService;
            this.injector = injector;
            this.baseCls = true;
            this.showDataId = true;
            this.headers = [];
            this.dataCodeFields = [];
            this.pageChanged = new i0.EventEmitter();
            this.pageSizeChanged = new i0.EventEmitter();
        }
        /**
         * @return {?}
         */
        ChgdrHeaderDatagridComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.resetColumnAndData(null, []);
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        ChgdrHeaderDatagridComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                var _this = this;
                if (!changes.chgdrs) {
                    //如果chgdrs没有更新，则直接返回
                    return;
                }
                /** @type {?} */
                var prevEntityId = this.getEntityId(changes.chgdrs.previousValue);
                /** @type {?} */
                var curEntityId = this.getEntityId(changes.chgdrs.currentValue);
                /** @type {?} */
                var requests = [
                    this.updateChgdrConfigHandler(curEntityId),
                    this.updateDataCodeFields(prevEntityId, curEntityId)
                ];
                rxjs.forkJoin(requests).subscribe(( /**
                 * @param {?} data
                 * @return {?}
                 */function (data) {
                    /** @type {?} */
                    var chgdrDatas = changes.chgdrs.currentValue;
                    if (!!chgdrDatas) {
                        _this.formatHeaders(chgdrDatas.headers);
                    }
                    //如果业务实体变更，则重置列信息
                    if (curEntityId != null && curEntityId != prevEntityId) {
                        _this.resetColumnAndData(curEntityId, chgdrDatas.headers);
                    }
                    else {
                        _this.headers = chgdrDatas.headers;
                    }
                }), ( /**
                 * @param {?} err
                 * @return {?}
                 */function (err) {
                    console.error("获取BE【" + curEntityId + "】的配置的业务变更日志的业务编号失败：", err);
                    _this.messageService.error(I18nUtil.instant("get-business-code-config-error", { errorMessage: ErrorUtil.getErrorMessage(err) }));
                }));
            };
        /** 更新业务实体元数据 */
        /**
         * 更新业务实体元数据
         * @private
         * @param {?} entityId
         * @return {?}
         */
        ChgdrHeaderDatagridComponent.prototype.updateChgdrConfigHandler = /**
         * 更新业务实体元数据
         * @private
         * @param {?} entityId
         * @return {?}
         */
            function (entityId) {
                var _this = this;
                if (entityId) {
                    return this.chgdrInnerService.getChgdrConfigHandler(entityId).pipe(operators.tap(( /**
                     * @param {?} data
                     * @return {?}
                     */function (data) { return _this.chgdrConfigHandler = data; })));
                }
                else {
                    return rxjs.of(this.chgdrConfigHandler);
                }
            };
        /**
         * 更新dataCodeField数据
         * @param prevEntityId 原BeId
         * @param curEntityId 当前BeId
         */
        /**
         * 更新dataCodeField数据
         * @param {?} prevEntityId 原BeId
         * @param {?} curEntityId 当前BeId
         * @return {?}
         */
        ChgdrHeaderDatagridComponent.prototype.updateDataCodeFields = /**
         * 更新dataCodeField数据
         * @param {?} prevEntityId 原BeId
         * @param {?} curEntityId 当前BeId
         * @return {?}
         */
            function (prevEntityId, curEntityId) {
                var _this = this;
                if (curEntityId != null && curEntityId != prevEntityId) {
                    //只有发生变更时才更新
                    return this.chgdrService.getRootEntityDataCodeFields(curEntityId).pipe(operators.tap(( /**
                     * @param {?} data
                     * @return {?}
                     */function (data) { return _this.dataCodeFields = data || []; })));
                }
                else {
                    //否则直接返回当前值
                    return rxjs.of(this.dataCodeFields);
                }
            };
        /**
         * 格式化headers。主要是解析其中的dataCode，将其转为Object
         * 使用dataCodeObj存放dataCode解析后的结果
         * dataCodeObj属性在ChangeDataHeader中不存在，此处通过强制赋值使用，并且只限于该处使用。
         */
        /**
         * 格式化headers。主要是解析其中的dataCode，将其转为Object
         * 使用dataCodeObj存放dataCode解析后的结果
         * dataCodeObj属性在ChangeDataHeader中不存在，此处通过强制赋值使用，并且只限于该处使用。
         * @private
         * @param {?} headers
         * @return {?}
         */
        ChgdrHeaderDatagridComponent.prototype.formatHeaders = /**
         * 格式化headers。主要是解析其中的dataCode，将其转为Object
         * 使用dataCodeObj存放dataCode解析后的结果
         * dataCodeObj属性在ChangeDataHeader中不存在，此处通过强制赋值使用，并且只限于该处使用。
         * @private
         * @param {?} headers
         * @return {?}
         */
            function (headers) {
                var _this = this;
                if (!headers) {
                    return;
                }
                headers.forEach(( /**
                 * @param {?} header
                 * @return {?}
                 */function (header) {
                    /** @type {?} */
                    var dataCodeObj;
                    _this.formatDataCodeMap(header.dataCode);
                    if (header.dataCode && header.dataCode.size == 1 && header.dataCode.has("_$dataCode")) {
                        //兼容老数据
                        //老数据dataCode为自由文本
                        dataCodeObj = {};
                        if (_this.dataCodeFields && _this.dataCodeFields.length > 0) {
                            dataCodeObj[_this.dataCodeFields[0].fieldLabelId] = header.dataCode.get("_$dataCode");
                        }
                    }
                    else {
                        dataCodeObj = MapUtil.convertMapToObject(header.dataCode);
                    }
                    (( /** @type {?} */(header))).dataCodeObj = dataCodeObj;
                }));
            };
        /** 格式化业务主键Map */
        /**
         * 格式化业务主键Map
         * @private
         * @param {?} dataCode
         * @return {?}
         */
        ChgdrHeaderDatagridComponent.prototype.formatDataCodeMap = /**
         * 格式化业务主键Map
         * @private
         * @param {?} dataCode
         * @return {?}
         */
            function (dataCode) {
                var _this = this;
                if (!dataCode || !this.chgdrConfigHandler) {
                    return;
                }
                Array.from(dataCode.keys()).forEach(( /**
                 * @param {?} fieldLabelId
                 * @return {?}
                 */function (fieldLabelId) {
                    /** @type {?} */
                    var value = dataCode.get(fieldLabelId);
                    /** @type {?} */
                    var formatedValue = _this.chgdrConfigHandler.formatMainObjectFieldValue(value, fieldLabelId);
                    dataCode.set(fieldLabelId, formatedValue);
                }));
            };
        /**
         * @private
         * @param {?} chgdrs
         * @return {?}
         */
        ChgdrHeaderDatagridComponent.prototype.getEntityId = /**
         * @private
         * @param {?} chgdrs
         * @return {?}
         */
            function (chgdrs) {
                if (!chgdrs || !chgdrs.headers || chgdrs.headers.length == 0) {
                    return null;
                }
                else {
                    return chgdrs.headers[0].entityId;
                }
            };
        /**
         * @private
         * @param {?} entityId
         * @param {?} headers
         * @return {?}
         */
        ChgdrHeaderDatagridComponent.prototype.resetColumnAndData = /**
         * @private
         * @param {?} entityId
         * @param {?} headers
         * @return {?}
         */
            function (entityId, headers) {
                /** @type {?} */
                var columns = [
                    { field: 'userName', width: 130, title: I18nUtil.instant("user-name-column-title") },
                    { field: 'changeTime', width: 130, title: I18nUtil.instant("change-time-column-title"), formatter: { type: 'datetime', options: { format: I18nUtil.instant("change-time-column-format") } } },
                    { field: 'operateType.name', width: 60, title: I18nUtil.instant("operation-type-title") },
                    { field: 'dataId', width: 130, title: I18nUtil.instant("tech-pk-column-title"), visible: this.showDataId }
                ];
                //{field: 'dataCode', width: 130, title: '业务编号'},
                // {field: 'reason', width: 130, title: '变更原因'},
                this.dataCodeFields.forEach(( /**
                 * @param {?} field
                 * @return {?}
                 */function (field) {
                    /** @type {?} */
                    var column = {
                        field: "dataCodeObj." + field.fieldLabelId,
                        title: field.fieldName,
                        width: 130
                    };
                    columns.push(column);
                }));
                /** @type {?} */
                var opColumn = { title: I18nUtil.instant("operation-column-title"), width: 130, template: this.opCell, halign: 'center', align: 'center' };
                columns.push(opColumn);
                this.columns = columns;
                this.headers = headers;
            };
        /**
         * @param {?} $event
         * @return {?}
         */
        ChgdrHeaderDatagridComponent.prototype.onPageChanged = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                this.pageChanged.emit($event);
            };
        /**
         * @param {?} $event
         * @return {?}
         */
        ChgdrHeaderDatagridComponent.prototype.onPageSizeChanged = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                this.pageSizeChanged.emit($event);
            };
        /**
         * @param {?} ctx
         * @return {?}
         */
        ChgdrHeaderDatagridComponent.prototype.clickShowChangeDetail = /**
         * @param {?} ctx
         * @return {?}
         */
            function (ctx) {
                var _this = this;
                /** @type {?} */
                var header = ctx.rowData;
                if (!header) {
                    return;
                }
                if (!this.beforeShowChangeDetail) {
                    //如果没有打开详情前回调，则按照默认方式打开
                    this.showChangeDetail(header);
                }
                else {
                    /** @type {?} */
                    var event_1 = new BeforeShowChangeDetailEvent();
                    event_1.header = header;
                    this.beforeShowChangeDetail(event_1).subscribe(( /**
                     * @param {?} isShow
                     * @return {?}
                     */function (isShow) {
                        if (isShow) {
                            _this.showChangeDetail(header);
                        }
                    }), ( /**
                     * @param {?} err
                     * @return {?}
                     */function (err) {
                        console.error("执行业务变更详情打开前事件报错：", err);
                        _this.messageService.error("执行业务变更详情打开前事件报错：" + ErrorUtil.getErrorMessage(err));
                    }));
                }
            };
        /**
         * @private
         * @param {?} header
         * @return {?}
         */
        ChgdrHeaderDatagridComponent.prototype.showChangeDetail = /**
         * @private
         * @param {?} header
         * @return {?}
         */
            function (header) {
                var _this = this;
                this.chgdrService.getChangeData(header.id, header.changeTime).subscribe(( /**
                 * @param {?} data
                 * @return {?}
                 */function (data) {
                    if (!data) {
                        _this.notifyService.error(I18nUtil.instant("can-not-found-change-record-by-id", { headerId: header.id }));
                    }
                    else {
                        _this.currentChgdrData = data;
                        _this.chgdrInfoDialog.show();
                    }
                }), ( /**
                 * @param {?} error
                 * @return {?}
                 */function (error) {
                    console.error("查询业务变更日志出错", error);
                    _this.messageService.error(I18nUtil.instant("query-change-data-record-error", { errorMessage: ErrorUtil.getErrorMessage(error) }));
                }));
            };
        Object.defineProperty(ChgdrHeaderDatagridComponent.prototype, "showChangeDetailBtn", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instant("show-change-detail-btn");
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ChgdrHeaderDatagridComponent.prototype, "changeDetailDialogTitle", {
            get: /**
             * @return {?}
             */ function () {
                return I18nUtil.instant("change-detail-dialog-title");
            },
            enumerable: true,
            configurable: true
        });
        ChgdrHeaderDatagridComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'lib-chgdr-header-datagrid',
                        template: "<farris-datagrid #chgdrGrid idField=\"id\" [columns]=\"columns\" [data]=\"headers\"\n  showLineNumber=\"true\" [fit]=\"true\" [fitColumns]=\"true\" \n  [pagination]=\"true\" [pageSize]=\"20\" [showPageList]=\"false\" [pageIndex]=\"chgdrs.pageIndex\" [total]=\"chgdrs.total\"\n  (pageChanged)=\"onPageChanged($event)\" (pageSizeChanged)=\"onPageSizeChanged($event)\"\n>\n</farris-datagrid>\n<ng-template #opCell let-ctx>\n  <a href=\"javascript: void(0);\" (click)=\"clickShowChangeDetail(ctx)\">{{showChangeDetailBtn}}</a>\n</ng-template>\n\n<!-- \u67E5\u770B\u53D8\u66F4\u65E5\u5FD7\u7684\u5F39\u7A97 -->\n<farris-dialog #chgdrInfoDialog\n  [title]=\"changeDetailDialogTitle\"\n  [width]=\"1100\"\n  [height]=\"600\"\n  [showMaxButton]=\"true\"\n  [resizable]=\"true\"\n  [showButtons]=\"false\"\n>\n  <chgdr-data-viewer [data]=\"currentChgdrData\"></chgdr-data-viewer>\n</farris-dialog>",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        ChgdrHeaderDatagridComponent.ctorParameters = function () {
            return [
                { type: ChgdrService },
                { type: i2$1.MessagerService },
                { type: uiNotify.NotifyService },
                { type: i0.ChangeDetectorRef },
                { type: ChgdrInnerService },
                { type: i0.Injector }
            ];
        };
        ChgdrHeaderDatagridComponent.propDecorators = {
            baseCls: [{ type: i0.HostBinding, args: ['class.f-utils-fill',] }],
            chgdrs: [{ type: i0.Input }],
            showDataId: [{ type: i0.Input }],
            beforeShowChangeDetail: [{ type: i0.Input }],
            pageChanged: [{ type: i0.Output }],
            pageSizeChanged: [{ type: i0.Output }],
            opCell: [{ type: i0.ViewChild, args: ['opCell',] }],
            chgdrInfoDialog: [{ type: i0.ViewChild, args: ["chgdrInfoDialog",] }]
        };
        return ChgdrHeaderDatagridComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/chgdr.module.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    common.registerLocaleData(zh);
    var ChgdrModule = /** @class */ (function () {
        function ChgdrModule() {
        }
        ChgdrModule.decorators = [
            { type: i0.NgModule, args: [{
                        declarations: [ChgdrDataViewerComponent, ChgdrListComponent, ChgdrRowBaseinfoFormComponent, ChgdrRowGridComponent, ChgdrListDialogContentComponent, ChgdrHeaderDatagridComponent],
                        imports: [
                            common.CommonModule,
                            forms.FormsModule,
                            core.TranslateModule.forRoot(),
                            uiSplitter.SplitterModule,
                            uiDraggable.AngularDraggableModule,
                            uiDatagrid.DatagridModule,
                            uiTreetable.TreeTableModule,
                            uiSection.FarrisSectionModule,
                            i2$1.MessagerModule.forRoot(),
                            uiNotify.NotifyModule,
                            uiDialog.FarrisDialogModule,
                            uiDatepicker.FarrisDatePickerModule,
                            i2.MetadataRTSelectModule,
                            i2.GSPMetadataRTServiceModule.forRoot('')
                        ],
                        providers: [
                            i2$1.MessagerService,
                            uiNotify.NotifyService,
                            i1.HttpService,
                            i1.SessionService,
                            i1.CacheService,
                            ChgdrListUiService,
                            ChgdrMetadataService,
                            ChgdrConfigService,
                            ChgdrInnerService,
                            ChgdrService,
                            I18nService
                        ],
                        entryComponents: [
                            ChgdrDataViewerComponent,
                            ChgdrListComponent,
                            ChgdrListDialogContentComponent
                        ],
                        exports: [ChgdrDataViewerComponent, ChgdrListComponent, ChgdrHeaderDatagridComponent]
                    },] }
        ];
        return ChgdrModule;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/model/change-data-header.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChangeDataHeader = /** @class */ (function () {
        function ChangeDataHeader() {
        }
        return ChangeDataHeader;
    }());
    var ChangeDataCompHeader = /** @class */ (function (_super) {
        __extends(ChangeDataCompHeader, _super);
        function ChangeDataCompHeader() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return ChangeDataCompHeader;
    }(ChangeDataHeader));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/model/change-data-row.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ChangeDataRow = /** @class */ (function () {
        function ChangeDataRow() {
        }
        return ChangeDataRow;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/model/chgdr-item-query-param.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 业务变更日志详情数据查询参数
     */
    var /**
     * 业务变更日志详情数据查询参数
     */ ChgdrItemQueryParam = /** @class */ (function () {
        function ChgdrItemQueryParam() {
        }
        return ChgdrItemQueryParam;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: public-api.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: gsp-cmp-chgdr.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.ChgdrModule = ChgdrModule;
    exports.ChgdrService = ChgdrService;
    exports.ChgdrDataViewerComponent = ChgdrDataViewerComponent;
    exports.ChgdrHeaderDatagridComponent = ChgdrHeaderDatagridComponent;
    exports.ChgdrListComponent = ChgdrListComponent;
    exports.ChangeDataHeader = ChangeDataHeader;
    exports.ChangeDataCompHeader = ChangeDataCompHeader;
    exports.ChangeDataRow = ChangeDataRow;
    exports.OperateType = OperateType;
    exports.ChangeDataQueryParam = ChangeDataQueryParam;
    exports.ChangeDataQueryResult = ChangeDataQueryResult;
    exports.BeforeShowChangeDetailEvent = BeforeShowChangeDetailEvent;
    exports.ChgdrListDialogConfig = ChgdrListDialogConfig;
    exports.ChgdrItemQueryParam = ChgdrItemQueryParam;
    exports.ChgdrListUiService = ChgdrListUiService;
    exports.ɵf = ChgdrListDialogContentComponent;
    exports.ɵd = ChgdrRowBaseinfoFormComponent;
    exports.ɵe = ChgdrRowGridComponent;
    exports.ɵb = ChgdrConfigService;
    exports.ɵc = ChgdrInnerService;
    exports.ɵa = ChgdrMetadataService;
    exports.ɵh = I18nService;
    exports.ɵg = I18nUtil;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
