// @farris/mobile-vue-adapter/esm2015/lib/utils/format-uitl.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/format-uitl.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { StringUtil, NumberUtil, BoolUtil, EnumUtil, DateUtil, } from '@farris/mobile-devkit';
export default {
    /**
     * 枚举格式化
     */
    enum: (/**
     * @param {?} value
     * @param {?} options
     * @return {?}
     */
    function (value, options) {
        return EnumUtil.format(value, options);
    }),
    /**
     * 日期格式化
     */
    date: (/**
     * @param {?} value
     * @param {?} options
     * @return {?}
     */
    function (value, options) {
        return DateUtil.format(value, options);
    }),
    /**
     * 日期操作
     */
    dateOperation: (/**
     * @param {?} value
     * @param {?} options
     * @return {?}
     */
    function (value, options) {
        return DateUtil.dateOperation(value, options);
    }),
    /**
     * 数字格式化
     */
    number: (/**
     * @param {?} value
     * @param {?} options
     * @return {?}
     */
    function (value, options) {
        return NumberUtil.format(value, options);
    }),
    /**
     * 字符串格式化
     *
     */
    string: (/**
     * @param {?} value
     * @param {?} options
     * @return {?}
     */
    function (value, options) {
        return StringUtil.format(value, options);
    }),
    /**
     * 布尔值格式化
     */
    bool: (/**
     * @param {?} value
     * @param {?} options
     * @return {?}
     */
    function (value, options) {
        return BoolUtil.format(value, options);
    }),
};
