// @edp-pmf/grapheditor/dist/cjs/lib/model/ProcessCell.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProcessCell = void 0;
var ElementPropertyConfig_1 = require("./farris-property-config/ElementPropertyConfig");
var mxgraph_ts_1 = require("@edp-pmf/mxgraph-ts");
/**
 * 流程元素类。
 * 取名虽然Cell，但实际上不需要继承mxCell以及Cell类，仅持有流程与画布实例，可进行属性获取与修改操作。
 */
var ProcessCell = /** @class */ (function () {
    function ProcessCell() {
    }
    ProcessCell.prototype.getFarrisPropConfig = function () {
        var _a;
        var propertyData = {};
        var usual = new ElementPropertyConfig_1.ElementPropertyConfig();
        usual.categoryId = 'usual';
        usual.categoryName = mxgraph_ts_1.mxResources.get("Basic Info") || '基本信息';
        usual.tabId = 'basic';
        usual.tabName = mxgraph_ts_1.mxResources.get("Basic Property") || '基础属性';
        usual.properties = [
            { propertyID: 'Name', propertyName: mxgraph_ts_1.mxResources.get("Property Name") || '名称', propertyType: 'string', visible: !this.editorUi.multiLangEnabled },
            { propertyID: "NameLanguage", propertyName: mxgraph_ts_1.mxResources.get("Property Name") || '名称', propertyType: 'multiLanguage', visible: this.editorUi.multiLangEnabled },
            { propertyID: "Key", propertyName: mxgraph_ts_1.mxResources.get("Property Id") || '标识', propertyType: 'string', readonly: true },
        ];
        propertyData["Name"] = this.process.Name;
        // 没有NameLanguage时把Name作为中文简体值
        propertyData["NameLanguage"] = this.process.NameLanguage || (_a = {},
            _a[this.editorUi.langCode || "zh-CHS"] = this.process.Name,
            _a);
        propertyData["Key"] = this.process.Key;
        propertyData["Id"] = this.process.Id;
        propertyData["ClrTypeId"] = this.process.ClrTypeID;
        propertyData["ClrTypeID"] = this.process.ClrTypeID;
        return {
            propertyConfig: [usual],
            propertyData: propertyData
        };
    };
    ProcessCell.prototype.updateProps = function (obj) {
        this.process[obj.propertyID] = obj.propertyValue;
        if (obj.propertyID == "NameLanguage") {
            this.process.Name = this.process.NameLanguage && this.process.NameLanguage['zh-CHS'] || "";
        }
    };
    return ProcessCell;
}());
exports.ProcessCell = ProcessCell;
