// @farris/ui-locale/esm5/farris-ui-locale.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { EN_US, ZH_CN, COLLAPSE_DIRECTIVE_LOCALE, DATAGRID_LOCALE, FILTER_EDITOR_LOCALE, ENUM_EDITOR_LOCALE, LOADING_LOCALE, LOOKUP_LOCALE, MESSAGER_LOCALE, MULTI_SELECT_LOCALE, NOTIFY_LOCALE, PAGINATION_LOCALE, QUERY_CONDITION_LOCALE, QUERY_SOLUTION_LOCALE, RESPONSE_TOOLBAR_LOCALE, SECTION_LOCALE, SIDEBAR_LOCALE, SORT_EDITOR_LOCALE, TABS_LOCALE, TEXT_LOCALE, TREETABLE_LOCALE, AVATAR_LOCALE, LIST_FILTER_LOCALE, PROGRESS_STEP_LOCALE, LANGUAGE_LABEL_LOCALE, VERIFY_DETAIL_LOCALE, COLLAPSE_DIRECTIVE_LOCALE_ZHCHS, DATAGRID_LOCALE_ZHCHS, FILTER_EDITOR_LOCALE_ZHCHS, ENUM_EDITOR_LOCALE_ZHCHS, LOADING_LOCALE_ZHCHS, LOOKUP_LOCALE_ZHCHS, MESSAGER_LOCALE_ZHCHS, MULTI_SELECT_LOCALE_ZHCHS, NOTIFY_LOCALE_ZHCHS, PAGINATION_LOCALE_ZHCHS, QUERY_CONDITION_LOCALE_ZHCHS, QUERY_SOLUTION_LOCALE_ZHCHS, RESPONSE_TOOLBAR_LOCALE_ZHCHS, SECTION_LOCALE_ZHCHS, SIDEBAR_LOCALE_ZHCHS, SORT_EDITOR_LOCALE_ZHCHS, TABS_LOCALE_ZHCHS, TEXT_LOCALE_ZHCHS, TREETABLE_LOCALE_ZHCHS, AVATAR_LOCALE_ZHCHS, LIST_FILTER_LOCALE_ZHCHS, PROGRESS_STEP_LOCALE_ZHCHS, LANGUAGE_LABEL_LOCALE_ZHCHS, VERIFY_DETAIL_ZHCHS, FARRIS_LOCAL_CUSTOM_DATA, LocaleService, FarrisLocalePipe, LocaleModule, FARRIS_LOCALES } from './public-api';
export { BATCH_EDIT_DIALOG_LOCALE as ɵd } from './lib/locales/en-us/batch-edit-dialog';
export { COMBO_LOCALE as ɵa } from './lib/locales/en-us/combo';
export { DISCUSSION_GROUP_LOCALE as ɵg } from './lib/locales/en-us/discussion-group';
export { FILTER_PANEL_LOCALE as ɵi } from './lib/locales/en-us/filter-panel';
export { FOOTER_LOCALE as ɵf } from './lib/locales/en-us/footer';
export { LISTVIEW_LOCALE as ɵc } from './lib/locales/en-us/list-view';
export { LOOKUP_CONFIG_LOCALE as ɵk } from './lib/locales/en-us/lookup-config';
export { NUMERIC_LOCALE as ɵb } from './lib/locales/en-us/numeric';
export { PAGE_WALKER_LOCALE as ɵe } from './lib/locales/en-us/page-walker';
export { SCROLLSPY_LOCALE as ɵj } from './lib/locales/en-us/scrollspy';
export { TAG_LOCALE as ɵh } from './lib/locales/en-us/tag';
export { AVATAR_LOCALE_ZHCHT as ɵbt } from './lib/locales/zh-CHT/avatar';
export { BATCH_EDIT_DIALOG_LOCALE_ZHCHT as ɵby } from './lib/locales/zh-CHT/batch-edit-dialog';
export { COLLAPSE_DIRECTIVE_LOCALE_ZHCHT as ɵbs } from './lib/locales/zh-CHT/collapse';
export { COMBO_LOCALE as ɵx } from './lib/locales/zh-CHT/combo';
export { DATAGRID_LOCALE_ZHCHT as ɵy } from './lib/locales/zh-CHT/datagrid';
export { DISCUSSION_GROUP_LOCALE_ZHCHT as ɵcb } from './lib/locales/zh-CHT/discussion-group';
export { ENUM_EDITOR_LOCALE_ZHCHT as ɵba, FILTER_EDITOR_LOCALE_ZHCHT as ɵz } from './lib/locales/zh-CHT/filter-editor';
export { FILTER_PANEL_LOCALE_ZHCHT as ɵcd } from './lib/locales/zh-CHT/filter-panel';
export { FOOTER_LOCALE_ZHCHT as ɵca } from './lib/locales/zh-CHT/footer';
export { ZH_CHT as ɵw } from './lib/locales/zh-CHT/index';
export { LANGUAGE_LABEL_LOCALE_ZHCHT as ɵbw } from './lib/locales/zh-CHT/language-label';
export { LIST_FILTER_LOCALE_ZHCHT as ɵbu } from './lib/locales/zh-CHT/list-filter';
export { LISTVIEW_LOCALE_ZHCHT as ɵbk } from './lib/locales/zh-CHT/list-view';
export { LOADING_LOCALE_ZHCHT as ɵbc } from './lib/locales/zh-CHT/loading';
export { LOOKUP_LOCALE_ZHCHT as ɵbb } from './lib/locales/zh-CHT/lookup';
export { LOOKUP_CONFIG_LOCALE_ZHCHT as ɵcf } from './lib/locales/zh-CHT/lookup-config';
export { MESSAGER_LOCALE_ZHCHT as ɵbd } from './lib/locales/zh-CHT/messager';
export { MULTI_SELECT_LOCALE_ZHCHT as ɵbi } from './lib/locales/zh-CHT/multi-select';
export { NOTIFY_LOCALE_ZHCHT as ɵbe } from './lib/locales/zh-CHT/notify';
export { NUMERIC_LOCALE_ZHCHT as ɵbf } from './lib/locales/zh-CHT/numeric';
export { PAGE_WALKER_ZHCHT as ɵbz } from './lib/locales/zh-CHT/page-walker';
export { PAGINATION_LOCALE_ZHCHT as ɵbo } from './lib/locales/zh-CHT/pagination';
export { PROGRESS_STEP_LOCALE_ZHCHT as ɵbv } from './lib/locales/zh-CHT/progress-step';
export { QUERY_CONDITION_LOCALE_ZHCHT as ɵbq } from './lib/locales/zh-CHT/query-condition';
export { QUERY_SOLUTION_LOCALE_ZHCHT as ɵbr } from './lib/locales/zh-CHT/query-solution';
export { RESPONSE_TOOLBAR_LOCALE_ZHCHT as ɵbp } from './lib/locales/zh-CHT/response-toolbar';
export { SCROLLSPY_LOCALE_ZHCHT as ɵce } from './lib/locales/zh-CHT/scrollspy';
export { SECTION_LOCALE_ZHCHT as ɵbn } from './lib/locales/zh-CHT/section';
export { SIDEBAR_LOCALE_ZHCHT as ɵbm } from './lib/locales/zh-CHT/sidebar';
export { SORT_EDITOR_LOCALE_ZHCHT as ɵbg } from './lib/locales/zh-CHT/sort-editor';
export { TABS_LOCALE_ZHCHT as ɵbj } from './lib/locales/zh-CHT/tabs';
export { TAG_LOCALE_ZHCHT as ɵcc } from './lib/locales/zh-CHT/tag';
export { TEXT_LOCALE_ZHCHT as ɵbl } from './lib/locales/zh-CHT/text';
export { TREETABLE_LOCALE_ZHCHT as ɵbh } from './lib/locales/zh-CHT/treetable';
export { VERIFY_DETAIL_ZHCHT as ɵbx } from './lib/locales/zh-CHT/verify-detail';
export { BATCH_EDIT_DIALOG_LOCALE_ZHCHS as ɵo } from './lib/locales/zh-cn/batch-edit-dialog';
export { COMBO_LOCALE as ɵl } from './lib/locales/zh-cn/combo';
export { DISCUSSION_GROUP_LOCALE_ZHCHS as ɵr } from './lib/locales/zh-cn/discussion-group';
export { FILTER_PANEL_LOCALE_ZHCHS as ɵt } from './lib/locales/zh-cn/filter-panel';
export { FOOTER_LOCALE_ZHCHS as ɵq } from './lib/locales/zh-cn/footer';
export { LISTVIEW_LOCALE_ZHCHS as ɵn } from './lib/locales/zh-cn/list-view';
export { LOOKUP_CONFIG_LOCALE_ZHCHS as ɵv } from './lib/locales/zh-cn/lookup-config';
export { NUMERIC_LOCALE_ZHCHS as ɵm } from './lib/locales/zh-cn/numeric';
export { PAGE_WALKER_ZHCHS as ɵp } from './lib/locales/zh-cn/page-walker';
export { SCROLLSPY_LOCALE_ZHCHS as ɵu } from './lib/locales/zh-cn/scrollspy';
export { TAG_LOCALE_ZHCHS as ɵs } from './lib/locales/zh-cn/tag';
