// @farris/mobile-devkit/esm5/lib/state-machine/state_machine_context.js
import { VariableParseService } from '../variable/index';
/**
 * 状态机上下文
 */
var StateMachineContext = /** @class */ (function () {
    /**
     * 构造函数
     * @param stateMachine 状态机实例
     * @param initialState 初始状态
     */
    function StateMachineContext(stateMachine, initialState) {
        this.stateMachine = stateMachine;
        this.state = initialState.name;
    }
    /**
     * 初始化
     */
    StateMachineContext.prototype.init = function (frameContext) {
        this.viewModelContext = frameContext;
        this.parser = this.viewModelContext.injector.get(VariableParseService);
        this.stateMachineWatcher = this.stateMachine.stateMachineWatcher;
    };
    /**
     * 状态迁移
     * @param stateName 下一状态的名称
     */
    StateMachineContext.prototype.transitTo = function (stateName) {
        var nextState = this.stateMachine.states[stateName];
        if (nextState) {
            this.state = nextState.name;
            this.stateMachine.render();
        }
    };
    /**
     * 获取expression对应的UIState值
     * @param expression UIState表达式
     */
    StateMachineContext.prototype.getUIState = function (expression) {
        if (!expression) {
            return;
        }
        var viewModelContext = this.stateMachineWatcher.getViewModelContext(expression);
        if (!viewModelContext) {
            return;
        }
        this.stateMachineWatcher.subscribeUIStateChange(viewModelContext, expression);
        if (this.parser) {
            var value = this.parser.parse(expression, viewModelContext);
            if (value === null) {
                return null;
            }
            if (typeof value === 'object' && Object.keys(value).length === 0) {
                return null;
            }
            return value;
        }
        else {
            throw new Error('未初始化变量解析器。');
        }
    };
    /**
     * 获取数据的值
     */
    StateMachineContext.prototype.getData = function (expression) {
        if (!expression) {
            return;
        }
        var viewModelContext = this.stateMachineWatcher.getViewModelContext(expression);
        if (!viewModelContext) {
            return;
        }
        this.stateMachineWatcher.subscribeEntityChange(viewModelContext, expression);
        if (this.parser) {
            var value = this.parser.parse(expression, viewModelContext);
            if (value === null) {
                return null;
            }
            if (typeof value === 'object' && Object.keys(value).length === 0) {
                return null;
            }
            return value;
        }
        else {
            throw new Error('未初始化变量解析器。');
        }
    };
    return StateMachineContext;
}());
export { StateMachineContext };
