// @farris/ui-list-filter/esm5/lib/service/solution.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
var SolutionService = /** @class */ (function () {
    function SolutionService() {
        this.filterEditors = {};
    }
    /**
     * @return {?}
     */
    SolutionService.prototype.getGuid = /**
     * @return {?}
     */
    function () {
        return (this.s4() + this.s4() + "-" + this.s4() + "-" + this.s4() + "-" + this.s4() + "-" + this.s4() + this.s4() + this.s4());
    };
    /**
     * @private
     * @return {?}
     */
    SolutionService.prototype.s4 = /**
     * @private
     * @return {?}
     */
    function () {
        return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
    };
    /**
     * 根据字段获取Lable控件的Title
     * @param field
     */
    /**
     * 根据字段获取Lable控件的Title
     * @param {?} field
     * @return {?}
     */
    SolutionService.prototype.getConditionItemTitle = /**
     * 根据字段获取Lable控件的Title
     * @param {?} field
     * @return {?}
     */
    function (field) {
        if (field.id && field.id !== '') {
            return field.fieldName;
        }
    };
    /**
     * 数组克隆 也可使用JSON.Parse(JSON.Stringify([input])),前提类结构中没有借口
     * @param input
     */
    /**
     * 数组克隆 也可使用JSON.Parse(JSON.Stringify([input])),前提类结构中没有借口
     * @param {?} input
     * @return {?}
     */
    SolutionService.prototype.arrayClone = /**
     * 数组克隆 也可使用JSON.Parse(JSON.Stringify([input])),前提类结构中没有借口
     * @param {?} input
     * @return {?}
     */
    function (input) {
        /** @type {?} */
        var result = new Array();
        if (input && input.length > 0) {
            input.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                result.push(Object.assign({}, element));
            }));
        }
        return result;
    };
    /**
     * @param {?} fieldName
     * @param {?} cmpIns
     * @return {?}
     */
    SolutionService.prototype.addFilterEditorInstance = /**
     * @param {?} fieldName
     * @param {?} cmpIns
     * @return {?}
     */
    function (fieldName, cmpIns) {
        this.filterEditors[fieldName] = cmpIns;
    };
    SolutionService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SolutionService.ctorParameters = function () { return []; };
    return SolutionService;
}());
export { SolutionService };
if (false) {
    /** @type {?} */
    SolutionService.prototype.filterEditors;
}
