// @gsp-lcm/metadatart-selector/esm5/metadata/metadatart-select/components/metadatartlist/metadatartlist.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadatart-select/components/metadatartlist/metadatartlist.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { MetadataRTService } from '../../metadatart.service';
import { process } from '@progress/kendo-data-query';
var TemplatemetadataRTComponent = /** @class */ (function () {
    function TemplatemetadataRTComponent(metadataService) {
        this.metadataService = metadataService;
        this.nameSpace = '命名空间';
        this.name = '名称';
        this.fileName = '文件名';
        this.type = '类型';
        this.businessObject = '业务对象';
        // tslint:disable-next-line:member-ordering
        this.state = {
            skip: 0,
            take: 50,
            // Initial filter descriptor
            filter: {
                logic: 'and',
                filters: []
            }
        };
        // tslint:disable-next-line:member-ordering
        this.selectableSetting = { mode: 'single', checkboxOnly: false };
    }
    /**
     * @return {?}
     */
    TemplatemetadataRTComponent.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.data = [];
    };
    /**
     * @param {?} state
     * @return {?}
     */
    TemplatemetadataRTComponent.prototype.dataStateChange = /**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this.state = state;
        this.gridData = process(this.data, this.state);
    };
    /**
     * @return {?}
     */
    TemplatemetadataRTComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (localStorage.getItem('languageCode') === 'en') {
            this.nameSpace = 'NameSpace';
            this.name = 'Name';
            this.fileName = 'FileName';
            this.type = 'Type';
            this.businessObject = 'BusinessObject';
        }
        /** @type {?} */
        var metadatadto = [];
        if (this.boId || this.bizTypeId) {
            this.metadataService.getMetadataListByBizTypeIdForHelp(this.boId, this.bizTypeId, this.typeList).subscribe((/**
             * @param {?} templates
             * @return {?}
             */
            function (templates) {
                if (templates) {
                    templates.forEach((/**
                     * @param {?} item
                     * @return {?}
                     */
                    function (item) {
                        metadatadto.push(item.metadata);
                    }));
                }
                _this.data = metadatadto;
                _this.gridData = process(_this.data, _this.state);
            }));
        }
        else {
            this.metadataService.GetMetadataRTListForHelp(this.typeList).subscribe((/**
             * @param {?} templates
             * @return {?}
             */
            function (templates) {
                if (templates) {
                    templates.forEach((/**
                     * @param {?} item
                     * @return {?}
                     */
                    function (item) {
                        metadatadto.push(item.metadata);
                    }));
                }
                _this.data = metadatadto;
                _this.gridData = process(_this.data, _this.state);
            }));
        }
    };
    // 选中元数据后的事件处理
    // 选中元数据后的事件处理
    /**
     * @param {?} event
     * @return {?}
     */
    TemplatemetadataRTComponent.prototype.onSelectionChange = 
    // 选中元数据后的事件处理
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.selectedRows.length > 0) {
            this.metadataService.clearMetadataRT();
            /** @type {?} */
            var metadata = {
                ID: event.selectedRows[0].dataItem.id,
                RelativePath: event.selectedRows[0].dataItem.relativePath,
                FileName: event.selectedRows[0].dataItem.fileName
            };
            this.metadataService.addMetadataRT(metadata);
        }
    };
    TemplatemetadataRTComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'app-metadatalist',
                    template: "<kendo-grid [data]=\"gridData\" \r\n[selectable]=\"selectableSetting\" \r\n(selectionChange)=onSelectionChange($event)    \r\n[pageSize]=\"state.take\"\r\n[skip]=\"state.skip\"\r\n[sort]=\"state.sort\"\r\n[filter]=\"state.filter\"\r\n[sortable]=\"false\"\r\n[pageable]=\"true\"\r\n[filterable]=\"true\"\r\n[resizable]=\"true\"\r\n(dataStateChange)=dataStateChange($event)>\r\n<kendo-grid-column [headerStyle]=\"{'border-right': '1px solid #d9d9d9'}\" field=\"nameSpace\" title={{nameSpace}} width=\"210\">\r\n</kendo-grid-column>\r\n<kendo-grid-column [headerStyle]=\"{'border-right': '1px solid #d9d9d9'}\" field=\"name\" title={{name}} width=\"200\">\r\n</kendo-grid-column>\r\n<kendo-grid-column [headerStyle]=\"{'border-right': '1px solid #d9d9d9'}\" field=\"fileName\" title={{fileName}} width=\"200\">\r\n</kendo-grid-column>\r\n<kendo-grid-column [headerStyle]=\"{'border-right': '1px solid #d9d9d9'}\" field=\"type\" title={{type}} width=\"150\">\r\n</kendo-grid-column>\r\n<kendo-grid-column field=\"bizobjectID\" title={{businessObject}} width=\"150\">\r\n</kendo-grid-column>\r\n</kendo-grid>",
                    styles: [".templateSelector{width:100%;height:475px;border:.5px solid #ddd}.templateDescription{width:200px;height:475px;border:.5px solid #ddd}.templateInfo{display:flex}"]
                }] }
    ];
    /** @nocollapse */
    TemplatemetadataRTComponent.ctorParameters = function () { return [
        { type: MetadataRTService }
    ]; };
    TemplatemetadataRTComponent.propDecorators = {
        typeList: [{ type: Input }],
        boId: [{ type: Input }],
        bizTypeId: [{ type: Input }]
    };
    return TemplatemetadataRTComponent;
}());
export { TemplatemetadataRTComponent };
if (false) {
    /** @type {?} */
    TemplatemetadataRTComponent.prototype.typeList;
    /** @type {?} */
    TemplatemetadataRTComponent.prototype.boId;
    /** @type {?} */
    TemplatemetadataRTComponent.prototype.bizTypeId;
    /** @type {?} */
    TemplatemetadataRTComponent.prototype.data;
    /** @type {?} */
    TemplatemetadataRTComponent.prototype.nameSpace;
    /** @type {?} */
    TemplatemetadataRTComponent.prototype.name;
    /** @type {?} */
    TemplatemetadataRTComponent.prototype.fileName;
    /** @type {?} */
    TemplatemetadataRTComponent.prototype.type;
    /** @type {?} */
    TemplatemetadataRTComponent.prototype.businessObject;
    /** @type {?} */
    TemplatemetadataRTComponent.prototype.state;
    /** @type {?} */
    TemplatemetadataRTComponent.prototype.gridData;
    /** @type {?} */
    TemplatemetadataRTComponent.prototype.selectableSetting;
    /**
     * @type {?}
     * @private
     */
    TemplatemetadataRTComponent.prototype.metadataService;
}
