// @gspwidget/widget-devkit/esm5/lib/widget-title-bar/widget-title-bar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, EventEmitter, Input, Output, TemplateRef, ViewEncapsulation } from '@angular/core';
import { JumpTypes } from '../preset/preset.jump';
import { WidgetDevkitService } from '../widget-devkit.service';
var WidgetTitleBarComponent = /** @class */ (function () {
    function WidgetTitleBarComponent(devkitSvc) {
        this.devkitSvc = devkitSvc;
        this.icon = '';
        /**
         * 标题
         */
        this.showTitle = true;
        this.title = '';
        /**
         * 数字
         */
        this.showNumber = false;
        this.number = 0;
        /**
         * 设置按钮
         */
        this.showSetting = false;
        this.clickSetting = new EventEmitter();
        /**
         * 刷新按钮
         */
        this.showRefresh = false;
        this.clickRefresh = new EventEmitter();
        this.JumpTypes = JumpTypes;
    }
    /**
     * @return {?}
     */
    WidgetTitleBarComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    WidgetTitleBarComponent.prototype.handleClickJump = /**
     * @return {?}
     */
    function () {
        this.devkitSvc.jumpWithoutData(this.jumpConfig);
    };
    WidgetTitleBarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-widget-title-bar',
                    template: "\n<div class=\"widget-title-bar d-flex\">\n  <!-- left -->\n  <div class=\"left align-items-center\">\n    <i *ngIf=\"icon!=''\" class=\"material-icons\">{{icon}}</i>\n    <div *ngIf=\"showTitle\" class=\"title\">{{title}}</div>\n    <div *ngIf=\"showNumber\" class=\"number\">{{number}}</div>\n    <ng-container *ngTemplateOutlet=\"leftTemplate\"></ng-container>\n  </div>\n  <!-- right -->\n  <div class=\"right\">\n    <ng-container *ngTemplateOutlet=\"rightTemplate\"></ng-container>\n    <div class=\"tool-wrap d-flex\" [title]=\"(settingTitle || '') | translate\">\n      <img\n        class=\"setting\"\n        *ngIf=\"showSetting\"\n        (click)=\"clickSetting.emit()\"\n        src=\"assets/img/setting.svg\"\n        [alt]=\"settingTitle\">\n    </div>\n    <div\n      class=\"tool-wrap d-flex\"\n      *ngIf=\"showRefresh\"\n      (click)=\"clickRefresh.emit()\">\n      <i class=\"icon-refresh material-icons\" style=\"font-size: 20px;\">refresh</i>\n    </div>\n    <div\n      *ngIf=\"jumpConfig && jumpConfig.jumpType!==JumpTypes.None\"\n      (click)=\"handleClickJump()\"\n      class=\"tool-wrap d-flex\">\n      <i class=\"material-icons\">chevron_right</i>\n    </div>\n  </div>\n</div>",
                    encapsulation: ViewEncapsulation.None,
                    styles: ["@charset \"UTF-8\";.widget-title-bar{display:flex;padding:12px 20px 10px}.widget-title-bar .left{flex:1 1 auto;display:flex}.widget-title-bar .left .material-icons{font-size:18px;margin-right:8px;color:#3794ff;padding-top:1px}.widget-title-bar .left .title{font-weight:700;font-size:15px}.widget-title-bar .left .number{font-size:15px;margin-left:10px}.widget-title-bar .right{flex:1 1 auto;display:flex;justify-content:flex-end;align-items:center}.widget-title-bar .right .widget-select{margin-left:5px}.widget-title-bar .right .setting{width:16px;height:16px;transition:.3s;opacity:1!important}.widget-title-bar .right .setting:hover{transform:rotate(60deg)}.widget-title-bar .right .tool-wrap{width:20px;margin-left:10px;font-size:14px;color:#000;transition:.3s;cursor:pointer;opacity:.2}.widget-title-bar .right .tool-wrap:hover{opacity:.4}"]
                }] }
    ];
    /** @nocollapse */
    WidgetTitleBarComponent.ctorParameters = function () { return [
        { type: WidgetDevkitService }
    ]; };
    WidgetTitleBarComponent.propDecorators = {
        icon: [{ type: Input }],
        showTitle: [{ type: Input }],
        title: [{ type: Input }],
        showNumber: [{ type: Input }],
        number: [{ type: Input }],
        showSetting: [{ type: Input }],
        settingTitle: [{ type: Input }],
        clickSetting: [{ type: Output }],
        showRefresh: [{ type: Input }],
        clickRefresh: [{ type: Output }],
        jumpConfig: [{ type: Input }],
        jumpProperty: [{ type: Input }],
        data: [{ type: Input }],
        leftTemplate: [{ type: ContentChild, args: ['leftTemplate',] }],
        rightTemplate: [{ type: ContentChild, args: ['rightTemplate',] }]
    };
    return WidgetTitleBarComponent;
}());
export { WidgetTitleBarComponent };
if (false) {
    /** @type {?} */
    WidgetTitleBarComponent.prototype.icon;
    /**
     * 标题
     * @type {?}
     */
    WidgetTitleBarComponent.prototype.showTitle;
    /** @type {?} */
    WidgetTitleBarComponent.prototype.title;
    /**
     * 数字
     * @type {?}
     */
    WidgetTitleBarComponent.prototype.showNumber;
    /** @type {?} */
    WidgetTitleBarComponent.prototype.number;
    /**
     * 设置按钮
     * @type {?}
     */
    WidgetTitleBarComponent.prototype.showSetting;
    /** @type {?} */
    WidgetTitleBarComponent.prototype.settingTitle;
    /** @type {?} */
    WidgetTitleBarComponent.prototype.clickSetting;
    /**
     * 刷新按钮
     * @type {?}
     */
    WidgetTitleBarComponent.prototype.showRefresh;
    /** @type {?} */
    WidgetTitleBarComponent.prototype.clickRefresh;
    /**
     * 是否启用跳转，开启后元部件配置中可配置全局联查
     * @type {?}
     */
    WidgetTitleBarComponent.prototype.jumpConfig;
    /**
     * @deprecated 应改用 enableJump。全局联查配置
     * @type {?}
     */
    WidgetTitleBarComponent.prototype.jumpProperty;
    /**
     * @deprecated 应改用 enableJump。为了跳转时能取到数据的第一行，用来拼接跳转路由参数。
     * @type {?}
     */
    WidgetTitleBarComponent.prototype.data;
    /**
     * 传入模板
     * @type {?}
     */
    WidgetTitleBarComponent.prototype.leftTemplate;
    /** @type {?} */
    WidgetTitleBarComponent.prototype.rightTemplate;
    /** @type {?} */
    WidgetTitleBarComponent.prototype.JumpTypes;
    /**
     * @type {?}
     * @private
     */
    WidgetTitleBarComponent.prototype.devkitSvc;
}
