// @edp-pmf/mxgraph-ts/dist/esm/lib/IO.js
import { mxgraphInstance } from './mxgraphFactory';
/**
 * IO
 */
// export const mxCellCodec = mxgraphInstance.mxCellCodec;
// export const mxChildChangeCodec = mxgraphInstance.mxChildChangeCodec;
export var mxCodec = mxgraphInstance.mxCodec;
export var mxCodecRegistry = mxgraphInstance.mxCodecRegistry;
// export const mxDefaultKeyHandlerCodec = mxgraphInstance.mxDefaultKeyHandlerCodec;
// export const mxDefaultPopupMenuCodec = mxgraphInstance.mxDefaultPopupMenuCodec;
export var mxDefaultToolbarCodec = mxgraphInstance.mxDefaultToolbarCodec;
// export const mxEditorCodec = mxgraphInstance.mxEditorCodec;
export var mxGenericChangeCodec = mxgraphInstance.mxGenericChangeCodec;
// export const mxGraphCodec = mxgraphInstance.mxGraphCodec;
// export const mxGraphViewCodec = mxgraphInstance.mxGraphViewCodec;
// export const mxModelCodec = mxgraphInstance.mxModelCodec;
export var mxObjectCodec = mxgraphInstance.mxObjectCodec;
// export const mxRootChangeCodec = mxgraphInstance.mxRootChangeCodec;
export var mxStylesheetCodec = mxgraphInstance.mxStylesheetCodec;
// export const mxTerminalChangeCodec = mxgraphInstance.mxTerminalChangeCodec;
