// @gsp-wf/rtdevkit/esm2015/lib/submit/assign-participant/assign-participant.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, Optional, ViewChild } from '@angular/core';
import { UserService } from '../../runtime/services/user.service';
import { Subject } from 'rxjs';
import { NotifyService } from '@farris/ui-notify';
import { TaskService } from '../../runtime/services/task.service';
import { I18nService } from '../../runtime/services/i18n.service';
import { DataTableComponent } from '@farris/ui-datatable';
import { LoadingService } from '@farris/ui-loading';
export class AssignParticipantComponent {
    /**
     * @param {?} userSvc
     * @param {?} notify
     * @param {?} taskSvc
     * @param {?} loadingService
     * @param {?} i18nService
     */
    constructor(userSvc, notify, taskSvc, loadingService, i18nService) {
        this.userSvc = userSvc;
        this.notify = notify;
        this.taskSvc = taskSvc;
        this.loadingService = loadingService;
        this.i18nService = i18nService;
        this.cls = 'farris-main-area flex-column';
        this.pageList = [10, 20, 30, 50, 100];
        this.paginationOptions = {
            id: 'WF-Users-DataTable-Pagination',
            currentPage: 1,
            itemsPerPage: 10,
            pageList: this.pageList,
            totalItems: 0
        };
        this.subject = new Subject();
        this.withAllOrg = true;
        this.orgId = '';
        this.userListData = [];
        // 是否允许多选,默认多选
        this.singleSelect = false;
        this.searchValue = null;
        this.selectedUsers = [];
        this.columns = [
            { title: this.i18nService ? this.i18nService.getValue('static.columnName.code') : '编号', field: 'Code', width: 80 },
            { title: this.i18nService ? this.i18nService.getValue('static.columnName.name') : '姓名', field: 'Name', width: 90 },
            { title: this.i18nService ? this.i18nService.getValue('static.columnName.department') : '组织信息', field: 'OrgName', width: 180 }
        ];
    }
    /**
     * @param {?} assigneeInfos
     * @return {?}
     */
    set assigneeInfo(assigneeInfos) {
        this.assignees = assigneeInfos;
        this.assigneesPa = {
            procInstId: '',
            //后端用不到 先为空
            assigneeInfos: []
        };
        this.getUserList();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    getUserList() {
        this.singleSelect = !this.assignees[0].multiSelect;
        if (this.assignees[0].assigneeUsers && this.assignees[0].assigneeUsers.length) {
            this.getAssignedUsers();
        }
        else {
            if (this.assignees[0].allowGlobalAssign) {
                if (this.assignees[0].organizationScope && this.assignees[0].organizationScope.length) {
                    this.withAllOrg = this.assignees[0].organizationScope[0].includeChildren;
                    this.orgId = this.assignees[0].organizationScope[0].id;
                }
                this.getAllUserList();
            }
            else {
                this.getAssignedUsers();
            }
        }
    }
    /**
     * @return {?}
     */
    getAssignedUsers() {
        if (this.assignees.length > 0) {
            this.selectedUsers = this.assignees[0].assigneeUsers.map((/**
             * @param {?} user
             * @return {?}
             */
            user => {
                return {
                    ID: user.id,
                    Name: user.name,
                    Code: user.code,
                    OrgId: user.orgId,
                    OrgName: user.orgNamePath
                };
            })),
                this.userListData = this.selectedUsers;
        }
    }
    /**
     * @protected
     * @return {?}
     */
    getAllUserList() {
        /** @type {?} */
        const load = this.loadingService.show();
        this.userSvc.getUserList_Assign(this.searchValue, this.paginationOptions.currentPage, this.paginationOptions.itemsPerPage, this.withAllOrg, this.orgId)
            .subscribe((/**
         * @param {?} re
         * @return {?}
         */
        (re) => {
            load.close();
            this.paginationOptions.totalItems = re.totalCount;
            this.userListData = re.data.map((/**
             * @param {?} u
             * @return {?}
             */
            u => ({
                ID: u.id,
                Code: u.code,
                Name: u.name,
                OrgId: u.sysOrgId,
                OrgName: u.orgNamePath,
                Checked: false
            })));
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            load.close();
        }));
    }
    /**
     * @param {?} event
     * @return {?}
     */
    search(event) {
        this.searchValue = event.value;
        if (this.assignees[0].assigneeUsers && this.assignees[0].assigneeUsers.length) { //有候选人，在候选人中筛选
            if (this.searchValue) {
                this.userListData = this.selectedUsers.filter((/**
                 * @param {?} user
                 * @return {?}
                 */
                user => (user.Name) && user.Name.includes(this.searchValue) ||
                    (user.Code) && user.Code.includes(this.searchValue)));
            }
            else {
                this.getAssignedUsers();
            }
        }
        else { //无候选人，在全员中选择
            this.paginationOptions.currentPage = 1;
            this.getUserList();
        }
    }
    /**
     * @return {?}
     */
    btnCancelClick() {
        this.notify.info(this.i18nService ? this.i18nService.getValue('static.assignee.cancelAssign') : '已取消操作。');
        this.subject.next(null);
    }
    /**
     * @return {?}
     */
    btnOKClick() {
        this.selectedUsers = this.singleSelect ? [this.userTable.selections] : this.userTable.selections;
        if (!this.selectedUsers || this.selectedUsers.length < 1) {
            this.notify.warning(this.i18nService ? this.i18nService.getValue('static.assignee.noAssignee') : '请至少选择一个审批人。');
            return;
        }
        // 添加已选人员
        /** @type {?} */
        const assigneeInfo = this.assignees.shift();
        assigneeInfo.assigneeIds = this.selectedUsers.map((/**
         * @param {?} user
         * @return {?}
         */
        user => user.ID));
        this.assigneesPa.assigneeInfos.push(assigneeInfo);
        // 检查指定是否完成，如果完成提交，否则继续指派任务
        if (this.assignees.length > 0) {
            this.userListData = []; // clear source
            this.getUserList();
        }
        else {
            /** @type {?} */
            const load = this.loadingService.show();
            this.taskSvc.assignNextUsersByProcInstId(this.assigneesPa).subscribe((/**
             * @return {?}
             */
            () => {
                load.close();
                this.notify.success(this.i18nService ? this.i18nService.getValue('static.assignee.successAssign') : '操作成功。');
                this.subject.next(this.assigneesPa);
            }), (/**
             * @param {?} err
             * @return {?}
             */
            err => {
                this.notify.error(this.i18nService ? this.i18nService.getValue('static.assignee.errorAssign') : '操作失败，请联系管理员查看原因！');
                this.subject.error(this.i18nService ? this.i18nService.getValue('static.assignee.errorAssign') : '操作失败，请联系管理员查看原因！');
            }));
        }
    }
    /**
     * @return {?}
     */
    subscribe() {
        return this.subject.asObservable();
    }
    /**
     * @param {?} pageSize
     * @return {?}
     */
    onPageSizeChange(pageSize) {
        this.paginationOptions.itemsPerPage = pageSize;
        this.getUserList();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onPageChange(event) {
        this.paginationOptions.currentPage = event.pageIndex;
        this.getUserList();
    }
}
AssignParticipantComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf-assign-participant',
                template: "<div class=\"modal-body farris-overflow-hidden f-utils-flex-column\" style=\"background-color: white;\">\r\n  <div class=\"d-flex flex-column my-data-grid h-100\" style=\"padding: 0 14px;\">\r\n    <div class=\"flex-auto h-100\" style=\"position:relative; overflow: auto;\">\r\n      <farris-datatable #allUserTable [columns]=\"columns\" [data]=\"userListData | paginate: paginationOptions\"\r\n        [showFilterBar]='true' [singleSelect]=\"singleSelect\" [idField]=\"'ID'\" [remote]=\"'server'\"\r\n        (search)=\"search($event)\" [pagination]='false'></farris-datatable>\r\n    </div>\r\n    <div style=\"position:relative;\">\r\n      <pagination-controls [id]=\"paginationOptions.id\" [previousLabel]=\"' '\" [nextLabel]=\"' '\" [showPageList]=\"false\"\r\n        (pageChange)=\"onPageChange($event)\" (pageSizeChange)=\"onPageSizeChange($event)\">\r\n      </pagination-controls>\r\n    </div>\r\n  </div>\r\n</div>\r\n<div class=\"modal-footer flex-shrink-0\">\r\n  <button type=\"button\" class=\"btn btn-primary\" (click)=\"btnOKClick()\">{{('static.footOk' | translate)?('static.footOk' | translate):'\u786E\u5B9A' }}</button>\r\n  <button type=\"button\" class=\"btn btn-default\" (click)=\"btnCancelClick()\">{{('static.footCancel' | translate)?('static.footCancel' | translate):'\u53D6\u6D88' }}</button>\r\n</div>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
AssignParticipantComponent.ctorParameters = () => [
    { type: UserService },
    { type: NotifyService },
    { type: TaskService },
    { type: LoadingService },
    { type: I18nService, decorators: [{ type: Optional }] }
];
AssignParticipantComponent.propDecorators = {
    cls: [{ type: HostBinding, args: ['class',] }],
    assigneeInfo: [{ type: Input }],
    userTable: [{ type: ViewChild, args: ['allUserTable',] }]
};
if (false) {
    /** @type {?} */
    AssignParticipantComponent.prototype.cls;
    /**
     * @type {?}
     * @private
     */
    AssignParticipantComponent.prototype.pageList;
    /** @type {?} */
    AssignParticipantComponent.prototype.paginationOptions;
    /**
     * @type {?}
     * @private
     */
    AssignParticipantComponent.prototype.assignees;
    /**
     * @type {?}
     * @private
     */
    AssignParticipantComponent.prototype.assigneesPa;
    /**
     * @type {?}
     * @private
     */
    AssignParticipantComponent.prototype.allowGlobalAssign;
    /**
     * @type {?}
     * @private
     */
    AssignParticipantComponent.prototype.subject;
    /**
     * @type {?}
     * @private
     */
    AssignParticipantComponent.prototype.withAllOrg;
    /**
     * @type {?}
     * @private
     */
    AssignParticipantComponent.prototype.orgId;
    /** @type {?} */
    AssignParticipantComponent.prototype.userListData;
    /** @type {?} */
    AssignParticipantComponent.prototype.singleSelect;
    /** @type {?} */
    AssignParticipantComponent.prototype.searchValue;
    /** @type {?} */
    AssignParticipantComponent.prototype.selectedUsers;
    /** @type {?} */
    AssignParticipantComponent.prototype.userTable;
    /** @type {?} */
    AssignParticipantComponent.prototype.columns;
    /**
     * @type {?}
     * @protected
     */
    AssignParticipantComponent.prototype.userSvc;
    /**
     * @type {?}
     * @protected
     */
    AssignParticipantComponent.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    AssignParticipantComponent.prototype.taskSvc;
    /**
     * @type {?}
     * @private
     */
    AssignParticipantComponent.prototype.loadingService;
    /**
     * @type {?}
     * @protected
     */
    AssignParticipantComponent.prototype.i18nService;
}
