// @gsp-wf/wf-process-model/esm5/lib/model/AifCreationBillActivity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, BpmnModelJsonHelper, StartEvent } from "@edp-pmf/bpmn-model";
var AifCreationBillActivity = /** @class */ (function (_super) {
    tslib_1.__extends(AifCreationBillActivity, _super);
    function AifCreationBillActivity(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = AifCreationBillActivity.Clr_Type_ID;
        _this.aifProcessCode = null;
        _this.voNodePath = null;
        _this.isRelied = true;
        _this.aifProcessCode = null;
        _this.voNodePath = null;
        _this.isRelied = true;
        return _this;
    }
    /**
     * @return {?}
     */
    AifCreationBillActivity.prototype.ResetBeforeLoadFrom = /**
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.aifProcessCode = null;
        this.voNodePath = null;
        this.isRelied = true;
    };
    /**
     * @return {?}
     */
    AifCreationBillActivity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, 'aifProcessCode', this.aifProcessCode);
        BpmnModelHelper.SetAttributeValue(obj, 'voNodePath', this.voNodePath);
        BpmnModelHelper.SetAttributeValue(obj, 'isRelied', this.isRelied);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AifCreationBillActivity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.aifProcessCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "aifProcessCode");
        this.voNodePath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voNodePath");
        this.isRelied = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isRelied");
    };
    /**
     * @return {?}
     */
    AifCreationBillActivity.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new AifCreationBillActivity(this.Model);
    };
    /**
    * 模型校验
    * @param errors 模型校验错误信息
    */
    /**
     * 模型校验
     * @param {?} errors 模型校验错误信息
     * @return {?}
     */
    AifCreationBillActivity.prototype.Validate = /**
     * 模型校验
     * @param {?} errors 模型校验错误信息
     * @return {?}
     */
    function (errors) {
        // TODO: 需要添加自定义校验信息
        _super.prototype.Validate.call(this, errors);
    };
    /**
     * @return {?}
     */
    AifCreationBillActivity.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var creationAct = new AifCreationBillActivity(this.Model);
        creationAct.Name = this.Name;
        creationAct.aifProcessCode = null;
        if (this.aifProcessCode) {
            creationAct.aifProcessCode = this.aifProcessCode;
        }
        creationAct.voNodePath = null;
        if (this.voNodePath) {
            creationAct.voNodePath = this.voNodePath;
        }
        creationAct.isRelied = this.isRelied;
        return creationAct;
    };
    AifCreationBillActivity.Clr_Type_ID = "wf.AifCreationBillActivity";
    return AifCreationBillActivity;
}(StartEvent));
export { AifCreationBillActivity };
if (false) {
    /** @type {?} */
    AifCreationBillActivity.Clr_Type_ID;
    /** @type {?} */
    AifCreationBillActivity.prototype.ClrTypeID;
    /** @type {?} */
    AifCreationBillActivity.prototype.aifProcessCode;
    /** @type {?} */
    AifCreationBillActivity.prototype.voNodePath;
    /** @type {?} */
    AifCreationBillActivity.prototype.isRelied;
}
BpmnModelHelper.RegisterElementType(new AifCreationBillActivity(new BpmnModel()));
