// @gsp-wf/wf-process-model/esm5/lib/model/ApprovalPivot.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionMethod, ExtensionElement } from "@edp-pmf/bpmn-model";
import { Condition } from "./extension/condition-expression/Condition";
var ApprovalPivot = /** @class */ (function (_super) {
    tslib_1.__extends(ApprovalPivot, _super);
    function ApprovalPivot(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = ApprovalPivot.Clr_Type_ID;
        return _this;
    }
    /**
     * @return {?}
     */
    ApprovalPivot.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "Type", this.type);
        ExtensionMethod.SetAttributeValue(obj, "Title", this.title);
        ExtensionMethod.SetAttributeValue(obj, "Description", this.description);
        if (this.condition) {
            BpmnModelHelper.SetAttributeValue(obj, "Condition", this.condition.ConvertToJsonObject());
        }
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ApprovalPivot.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.type = (/** @type {?} */ (ExtensionMethod.GetAttributeValue4String(obj, "Type")));
        this.title = ExtensionMethod.GetAttributeValue4String(obj, "Title");
        this.description = ExtensionMethod.GetAttributeValue4String(obj, "Description");
        /** @type {?} */
        var conditionElement = ExtensionMethod.GetAttributeValue(obj, 'Condition');
        if (conditionElement) {
            /** @type {?} */
            var condition = new Condition(this.Model);
            condition.LoadFromJsonObject(conditionElement);
            this.condition = condition;
        }
    };
    /**
     * @return {?}
     */
    ApprovalPivot.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new ApprovalPivot(this.Model);
    };
    /**
     * @return {?}
     */
    ApprovalPivot.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var approvalPivot = new ApprovalPivot(this.Model);
        approvalPivot.type = this.type;
        approvalPivot.title = this.title;
        approvalPivot.description = this.description;
        if (this.condition) {
            approvalPivot.condition = this.condition.clone();
        }
        return approvalPivot;
    };
    ApprovalPivot.Clr_Type_ID = "wf.ApprovalPivot";
    return ApprovalPivot;
}(ExtensionElement));
export { ApprovalPivot };
if (false) {
    /** @type {?} */
    ApprovalPivot.Clr_Type_ID;
    /** @type {?} */
    ApprovalPivot.prototype.ClrTypeID;
    /** @type {?} */
    ApprovalPivot.prototype.type;
    /** @type {?} */
    ApprovalPivot.prototype.condition;
    /** @type {?} */
    ApprovalPivot.prototype.title;
    /** @type {?} */
    ApprovalPivot.prototype.description;
}
BpmnModelHelper.RegisterElementType(new ApprovalPivot(new BpmnModel()));
