// @gsp-wf/wf-process-model/esm5/lib/model/ApprovalPivotConfig.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionMethod, ExtensionElement } from "@edp-pmf/bpmn-model";
import { ApprovalPivot } from "./ApprovalPivot";
var ApprovalPivotConfig = /** @class */ (function (_super) {
    tslib_1.__extends(ApprovalPivotConfig, _super);
    function ApprovalPivotConfig(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = ApprovalPivotConfig.Clr_Type_ID;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    ApprovalPivotConfig.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.approvalPivots = new Array();
    };
    /**
     * @return {?}
     */
    ApprovalPivotConfig.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "Title", this.title);
        ExtensionMethod.SetAttributeValue(obj, "Enable", this.enable);
        ExtensionMethod.SetAttributeValue(obj, "Expand", this.expand);
        /** @type {?} */
        var approvalPivots = new Array();
        if (this.approvalPivots !== null && this.approvalPivots.length > 0) {
            this.approvalPivots.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                approvalPivots.push(item.ConvertToJsonObject());
            }));
        }
        ExtensionMethod.SetAttributeValue(obj, "ApprovalPivots", approvalPivots);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ApprovalPivotConfig.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.title = ExtensionMethod.GetAttributeValue4String(obj, "Title");
        this.enable = ExtensionMethod.GetAttributeValue4Boolean(obj, "Enable");
        this.expand = ExtensionMethod.GetAttributeValue4Boolean(obj, "Expand");
        /** @type {?} */
        var approvalPivots = (/** @type {?} */ (ExtensionMethod.GetAttributeValue(obj, "ApprovalPivots")));
        if (approvalPivots !== null && approvalPivots.length > 0) {
            approvalPivots.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                /** @type {?} */
                var approvalPivot = new ApprovalPivot(_this.Model);
                approvalPivot.LoadFromJsonObject(item);
                _this.approvalPivots.push(approvalPivot);
            }));
        }
    };
    /**
     * @return {?}
     */
    ApprovalPivotConfig.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new ApprovalPivotConfig(this.Model);
    };
    /**
     * @return {?}
     */
    ApprovalPivotConfig.prototype.clone = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        /** @type {?} */
        var approvalPivotConfig = new ApprovalPivotConfig(this.Model);
        approvalPivotConfig.title = this.title;
        approvalPivotConfig.enable = this.enable;
        approvalPivotConfig.expand = this.expand;
        approvalPivotConfig.approvalPivots = new Array();
        if (this.approvalPivots && this.approvalPivots.length > 0) {
            try {
                for (var _b = tslib_1.__values(this.approvalPivots), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var p = _c.value;
                    /** @type {?} */
                    var approvalPivot = p.clone();
                    approvalPivotConfig.approvalPivots.push(approvalPivot);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        return approvalPivotConfig;
    };
    ApprovalPivotConfig.Clr_Type_ID = "wf.ApprovalPivotConfig";
    return ApprovalPivotConfig;
}(ExtensionElement));
export { ApprovalPivotConfig };
if (false) {
    /** @type {?} */
    ApprovalPivotConfig.Clr_Type_ID;
    /** @type {?} */
    ApprovalPivotConfig.prototype.ClrTypeID;
    /** @type {?} */
    ApprovalPivotConfig.prototype.title;
    /** @type {?} */
    ApprovalPivotConfig.prototype.enable;
    /** @type {?} */
    ApprovalPivotConfig.prototype.expand;
    /** @type {?} */
    ApprovalPivotConfig.prototype.approvalPivots;
}
BpmnModelHelper.RegisterElementType(new ApprovalPivotConfig(new BpmnModel()));
