// @gsp-wf/wf-process-model/esm5/lib/model/AutoHandleExpireOperation.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { ExpireOperation } from "./ExpireOperation";
var AutoHandleExpireOperation = /** @class */ (function (_super) {
    tslib_1.__extends(AutoHandleExpireOperation, _super);
    function AutoHandleExpireOperation() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.ClrTypeID = AutoHandleExpireOperation.Clr_Type_ID;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    AutoHandleExpireOperation.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    AutoHandleExpireOperation.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "Message", this.message);
        ExtensionMethod.SetAttributeValue(obj, "AutoHandleType", this.autoHandleType);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AutoHandleExpireOperation.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.message = ExtensionMethod.GetAttributeValue4String(obj, "Message");
        this.autoHandleType = ExtensionMethod.GetAttributeValue4String(obj, "AutoHandleType");
    };
    /**
     * @return {?}
     */
    AutoHandleExpireOperation.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new AutoHandleExpireOperation(this.Model);
    };
    /**
     * @return {?}
     */
    AutoHandleExpireOperation.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var autoHandleExpireOperation = new AutoHandleExpireOperation(this.Model);
        autoHandleExpireOperation.id = BpmnModelHelper.GenerateElementId();
        autoHandleExpireOperation.durationType = this.durationType;
        autoHandleExpireOperation.timeDuration = this.timeDuration;
        autoHandleExpireOperation.timeUnit = this.timeUnit;
        autoHandleExpireOperation.message = this.message;
        autoHandleExpireOperation.autoHandleType = this.autoHandleType;
        return autoHandleExpireOperation;
    };
    AutoHandleExpireOperation.Clr_Type_ID = "wf.Mex.AutoHandleExpireOperation";
    return AutoHandleExpireOperation;
}(ExpireOperation));
export { AutoHandleExpireOperation };
if (false) {
    /** @type {?} */
    AutoHandleExpireOperation.Clr_Type_ID;
    /** @type {?} */
    AutoHandleExpireOperation.prototype.ClrTypeID;
    /**
     * 办理类型，通过驳回否决等等
     * @type {?}
     */
    AutoHandleExpireOperation.prototype.autoHandleType;
    /**
     * 办理意见
     * @type {?}
     */
    AutoHandleExpireOperation.prototype.message;
}
BpmnModelHelper.RegisterElementType(new AutoHandleExpireOperation(new BpmnModel()));
