// @gsp-wf/wf-process-model/esm5/lib/model/ExpireOperation.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ExtensionMethod, MexExtensionElement } from "@edp-pmf/bpmn-model";
/**
 * @abstract
 */
var /**
 * @abstract
 */
ExpireOperation = /** @class */ (function (_super) {
    tslib_1.__extends(ExpireOperation, _super);
    function ExpireOperation(model) {
        return _super.call(this, model) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    ExpireOperation.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    ExpireOperation.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "Id", this.id);
        ExtensionMethod.SetAttributeValue(obj, "DurationType", this.durationType);
        ExtensionMethod.SetAttributeValue(obj, "TimeDuration", this.timeDuration);
        ExtensionMethod.SetAttributeValue(obj, "TimeUnit", this.timeUnit);
        ExtensionMethod.SetAttributeValue(obj, "OperationTypeName", this.operationTypeName);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ExpireOperation.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.id = ExtensionMethod.GetAttributeValue4String(obj, "Id");
        this.durationType = (/** @type {?} */ (ExtensionMethod.GetAttributeValue4String(obj, "DurationType")));
        this.timeDuration = ExtensionMethod.GetAttributeValue4Int64(obj, "TimeDuration");
        this.timeUnit = ExtensionMethod.GetAttributeValue4String(obj, "TimeUnit");
        this.operationTypeName = ExtensionMethod.GetAttributeValue4String(obj, "OperationTypeName");
    };
    return ExpireOperation;
}(MexExtensionElement));
/**
 * @abstract
 */
export { ExpireOperation };
if (false) {
    /** @type {?} */
    ExpireOperation.prototype.id;
    /**
     * 期限控制方式
     * @type {?}
     */
    ExpireOperation.prototype.durationType;
    /**
     * 任务期限
     * @type {?}
     */
    ExpireOperation.prototype.timeDuration;
    /**
     * 时间单位
     * @type {?}
     */
    ExpireOperation.prototype.timeUnit;
    /**
     * 操作类型名称
     * @type {?}
     */
    ExpireOperation.prototype.operationTypeName;
}
