// @gsp-wf/wf-process-model/esm5/lib/model/ExtendProperty.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
var ExtendProperty = /** @class */ (function (_super) {
    tslib_1.__extends(ExtendProperty, _super);
    function ExtendProperty(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = ExtendProperty.Clr_Type_ID;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    ExtendProperty.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    ExtendProperty.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "Key", this.key);
        ExtensionMethod.SetAttributeValue(obj, "Value", this.value);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ExtendProperty.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.key = ExtensionMethod.GetAttributeValue4String(obj, "Key");
        this.value = ExtensionMethod.GetAttributeValue4String(obj, "Value");
    };
    /**
     * @return {?}
     */
    ExtendProperty.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new ExtendProperty(this.Model);
    };
    /**
     * @return {?}
     */
    ExtendProperty.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var extendProperty = new ExtendProperty(this.Model);
        extendProperty.key = this.key;
        extendProperty.value = this.value;
        return extendProperty;
    };
    ExtendProperty.Clr_Type_ID = "wf.ExtendProperty";
    return ExtendProperty;
}(ExtensionElement));
export { ExtendProperty };
if (false) {
    /** @type {?} */
    ExtendProperty.Clr_Type_ID;
    /** @type {?} */
    ExtendProperty.prototype.ClrTypeID;
    /**
     * 扩展属性key值
     * @type {?}
     */
    ExtendProperty.prototype.key;
    /**
     * 扩展属性value值
     * @type {?}
     */
    ExtendProperty.prototype.value;
}
BpmnModelHelper.RegisterElementType(new ExtendProperty(new BpmnModel()));
