// @gsp-wf/wf-process-model/esm5/lib/model/ImageSetting.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
var ImageSettings = /** @class */ (function (_super) {
    tslib_1.__extends(ImageSettings, _super);
    function ImageSettings(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = ImageSettings.Clr_Type_ID;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    ImageSettings.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    ImageSettings.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "ImageOperation", this.imageOperation);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ImageSettings.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.imageOperation = ExtensionMethod.GetAttributeValue4String(obj, "ImageOperation");
    };
    /**
     * @return {?}
     */
    ImageSettings.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new ImageSettings(this.Model);
    };
    /**
     * @return {?}
     */
    ImageSettings.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var imageSettings = new ImageSettings(this.Model);
        imageSettings.imageOperation = this.imageOperation;
        return imageSettings;
    };
    ImageSettings.Clr_Type_ID = "wf.ImageSettings";
    return ImageSettings;
}(ExtensionElement));
export { ImageSettings };
if (false) {
    /** @type {?} */
    ImageSettings.Clr_Type_ID;
    /** @type {?} */
    ImageSettings.prototype.ClrTypeID;
    /**
     * 影像操作权限
     * @type {?}
     */
    ImageSettings.prototype.imageOperation;
}
BpmnModelHelper.RegisterElementType(new ImageSettings(new BpmnModel()));
