// @gsp-wf/wf-process-model/esm5/lib/model/OrganizationPotentialOwner.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, PotentialOwner } from "@edp-pmf/bpmn-model";
import { ActionScope } from "./ActionScope";
import { OrganizationLevel } from "./OrganizationLevel";
var OrganizationPotentialOwner = /** @class */ (function (_super) {
    tslib_1.__extends(OrganizationPotentialOwner, _super);
    function OrganizationPotentialOwner(model, id, name) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = OrganizationPotentialOwner.Clr_Type_ID;
        _this.MEXTAG = 'wf.OrganizationPotentialOwner';
        _this.Id = id;
        _this.Name = name;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    OrganizationPotentialOwner.prototype.resetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.actionScope = ActionScope.AssignUser;
        this.organizationLevel = OrganizationLevel.DEPARTMENT;
    };
    /**
     * @return {?}
     */
    OrganizationPotentialOwner.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "ActionScope", this.actionScope);
        BpmnModelHelper.SetAttributeValue(obj, "OrganizationLevel", this.organizationLevel);
        BpmnModelHelper.SetAttributeValue(obj, "OrganizationId", this.organizationId);
        BpmnModelHelper.SetAttributeValue(obj, "OrganizationCode", this.organizationCode);
        BpmnModelHelper.SetAttributeValue(obj, "OrganizationName", this.organizationName);
        BpmnModelHelper.SetAttributeValue(obj, "IncludeChildren", this.includeChildren);
        BpmnModelHelper.SetAttributeValue(obj, "ParentsUntil", this.parentsUntil);
        BpmnModelHelper.SetAttributeValue(obj, "Tag", this.tag);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    OrganizationPotentialOwner.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.actionScope = obj.ActionScope;
        this.organizationLevel = obj.OrganizationLevel;
        this.organizationId = obj.OrganizationId;
        this.organizationCode = obj.OrganizationCode;
        this.organizationName = obj.OrganizationName;
        this.includeChildren = obj.IncludeChildren;
        this.parentsUntil = obj.ParentsUntil;
        this.tag = obj.Tag;
    };
    /**
     * @protected
     * @return {?}
     */
    OrganizationPotentialOwner.prototype.GetExtensionElements4ConvertTo = /**
     * @protected
     * @return {?}
     */
    function () {
        /** @type {?} */
        var list = _super.prototype.GetExtensionElements4ConvertTo.call(this);
        return list;
    };
    /**
     * @protected
     * @param {?} extensionElements
     * @return {?}
     */
    OrganizationPotentialOwner.prototype.SetExtensionElements4LoadFrom = /**
     * @protected
     * @param {?} extensionElements
     * @return {?}
     */
    function (extensionElements) {
        _super.prototype.SetExtensionElements4LoadFrom.call(this, extensionElements);
    };
    /**
     * @return {?}
     */
    OrganizationPotentialOwner.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new OrganizationPotentialOwner(this.Model);
    };
    /**
     * @return {?}
     */
    OrganizationPotentialOwner.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var organizationPotentialOwner = new OrganizationPotentialOwner(this.Model);
        organizationPotentialOwner.Id = BpmnModelHelper.GenerateElementId();
        organizationPotentialOwner.actionScope = this.actionScope;
        organizationPotentialOwner.organizationLevel = this.organizationLevel;
        organizationPotentialOwner.organizationId = this.organizationId;
        organizationPotentialOwner.organizationName = this.organizationName;
        organizationPotentialOwner.organizationCode = this.organizationCode;
        organizationPotentialOwner.includeChildren = this.includeChildren;
        organizationPotentialOwner.parentsUntil = this.parentsUntil;
        organizationPotentialOwner.tag = this.tag;
        return organizationPotentialOwner;
    };
    OrganizationPotentialOwner.Clr_Type_ID = 'wf.OrganizationPotentialOwner';
    return OrganizationPotentialOwner;
}(PotentialOwner));
export { OrganizationPotentialOwner };
if (false) {
    /** @type {?} */
    OrganizationPotentialOwner.Clr_Type_ID;
    /** @type {?} */
    OrganizationPotentialOwner.prototype.ClrTypeID;
    /** @type {?} */
    OrganizationPotentialOwner.prototype.MEXTAG;
    /**
     * 作用域：AssignUser, Transfer, AddSignFront, AddSignBehind
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.actionScope;
    /**
     * 部门：DEPARTMENT；单位：UNIT
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.organizationLevel;
    /**
     * 组织Id：currentDepartment, currentUnit
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.organizationId;
    /**
     * 组织编号
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.organizationCode;
    /**
     * 组织名称
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.organizationName;
    /**
     * 包含下级
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.includeChildren;
    /**
     * 遍历父级
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.parentsUntil;
    /**
     * variable 变量值
     * expression 表达式值
     * constant 常量值
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.tag;
    /**
     * 暂存人员信息
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.participantInfo;
}
BpmnModelHelper.RegisterElementType(new OrganizationPotentialOwner(new BpmnModel()));
