// @gsp-wf/wf-process-model/esm5/lib/model/SignalEventDefinitionImpl.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionMethod, SignalEventDefinition } from "@edp-pmf/bpmn-model";
var SignalEventDefinitionImpl = /** @class */ (function (_super) {
    tslib_1.__extends(SignalEventDefinitionImpl, _super);
    function SignalEventDefinitionImpl(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = SignalEventDefinitionImpl.Clr_Type_ID;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    SignalEventDefinitionImpl.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    SignalEventDefinitionImpl.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "Name", this.name);
        ExtensionMethod.SetAttributeValue(obj, "TriggerAction", this.triggerAction);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    SignalEventDefinitionImpl.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.name = ExtensionMethod.GetAttributeValue4String(obj, "Name");
        this.triggerAction = ExtensionMethod.GetAttributeValue4String(obj, "TriggerAction");
    };
    /**
     * @return {?}
     */
    SignalEventDefinitionImpl.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new SignalEventDefinitionImpl(this.Model);
    };
    /**
     * @return {?}
     */
    SignalEventDefinitionImpl.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var signalEventDefinitionImpl = new SignalEventDefinitionImpl(this.Model);
        signalEventDefinitionImpl.name = this.name;
        signalEventDefinitionImpl.triggerAction = this.triggerAction;
        signalEventDefinitionImpl.SignalExpression = this.SignalExpression;
        signalEventDefinitionImpl.SignalRef = this.SignalRef;
        signalEventDefinitionImpl.Async = this.Async;
        return signalEventDefinitionImpl;
    };
    SignalEventDefinitionImpl.Clr_Type_ID = "wf.SignalEventDefinitionImpl";
    return SignalEventDefinitionImpl;
}(SignalEventDefinition));
export { SignalEventDefinitionImpl };
if (false) {
    /** @type {?} */
    SignalEventDefinitionImpl.Clr_Type_ID;
    /** @type {?} */
    SignalEventDefinitionImpl.prototype.ClrTypeID;
    /** @type {?} */
    SignalEventDefinitionImpl.prototype.name;
    /** @type {?} */
    SignalEventDefinitionImpl.prototype.triggerAction;
}
BpmnModelHelper.RegisterElementType(new SignalEventDefinitionImpl(new BpmnModel()));
