// @gsp-wf/wf-process-model/esm5/lib/model/SmsMessageTemplate.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { SmsMessageVariable } from "./SmsMessageVariable";
var SmsMessageTemplate = /** @class */ (function (_super) {
    tslib_1.__extends(SmsMessageTemplate, _super);
    function SmsMessageTemplate(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = SmsMessageTemplate.Clr_Type_ID;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    SmsMessageTemplate.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.smsMessageVariables = new Array();
    };
    /**
     * @return {?}
     */
    SmsMessageTemplate.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "Code", this.code);
        ExtensionMethod.SetAttributeValue(obj, "Content", this.content);
        if (this.smsMessageVariables && this.smsMessageVariables.length > 0) {
            /** @type {?} */
            var variables = new Array();
            try {
                for (var _b = tslib_1.__values(this.smsMessageVariables), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var variable = _c.value;
                    variables.push(variable.ConvertToJsonObject());
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            BpmnModelHelper.SetAttributeValue(obj, "SmsMessageVariables", variables);
        }
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    SmsMessageTemplate.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var e_2, _a;
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.code = ExtensionMethod.GetAttributeValue4String(obj, "Code");
        this.content = ExtensionMethod.GetAttributeValue4String(obj, "Content");
        this.smsMessageVariables = new Array();
        /** @type {?} */
        var variables = ExtensionMethod.GetAttributeValue(obj, "SmsMessageVariables");
        if (variables && variables.length > 0) {
            try {
                for (var variables_1 = tslib_1.__values(variables), variables_1_1 = variables_1.next(); !variables_1_1.done; variables_1_1 = variables_1.next()) {
                    var variable = variables_1_1.value;
                    /** @type {?} */
                    var smsMessageVariable = new SmsMessageVariable(this.Model);
                    smsMessageVariable.LoadFromJsonObject(variable);
                    this.smsMessageVariables.push(smsMessageVariable);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (variables_1_1 && !variables_1_1.done && (_a = variables_1.return)) _a.call(variables_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
    };
    /**
     * @return {?}
     */
    SmsMessageTemplate.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new SmsMessageTemplate(this.Model);
    };
    /**
     * @return {?}
     */
    SmsMessageTemplate.prototype.clone = /**
     * @return {?}
     */
    function () {
        var e_3, _a;
        /** @type {?} */
        var smsMessageTemplate = new SmsMessageTemplate(this.Model);
        smsMessageTemplate.code = this.code;
        smsMessageTemplate.content = this.content;
        smsMessageTemplate.smsMessageVariables = new Array();
        if (this.smsMessageVariables && this.smsMessageVariables.length > 0) {
            try {
                for (var _b = tslib_1.__values(this.smsMessageVariables), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var variable = _c.value;
                    /** @type {?} */
                    var smsMessageVariable = variable.clone();
                    smsMessageTemplate.smsMessageVariables.push(smsMessageVariable);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        return smsMessageTemplate;
    };
    SmsMessageTemplate.Clr_Type_ID = "wf.SmsMessageTemplate";
    return SmsMessageTemplate;
}(ExtensionElement));
export { SmsMessageTemplate };
if (false) {
    /** @type {?} */
    SmsMessageTemplate.Clr_Type_ID;
    /** @type {?} */
    SmsMessageTemplate.prototype.ClrTypeID;
    /** @type {?} */
    SmsMessageTemplate.prototype.code;
    /** @type {?} */
    SmsMessageTemplate.prototype.content;
    /** @type {?} */
    SmsMessageTemplate.prototype.smsMessageVariables;
}
BpmnModelHelper.RegisterElementType(new SmsMessageTemplate(new BpmnModel()));
