// @gsp-wf/wf-process-model/esm5/lib/model/SmsMessageVariable.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
var SmsMessageVariable = /** @class */ (function (_super) {
    tslib_1.__extends(SmsMessageVariable, _super);
    function SmsMessageVariable(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = SmsMessageVariable.Clr_Type_ID;
        _this.tag = 'expression';
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    SmsMessageVariable.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    SmsMessageVariable.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "Code", this.code);
        ExtensionMethod.SetAttributeValue(obj, "Value", this.value);
        ExtensionMethod.SetAttributeValue(obj, "Tag", this.tag);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    SmsMessageVariable.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.code = ExtensionMethod.GetAttributeValue4String(obj, "Code");
        this.value = ExtensionMethod.GetAttributeValue4String(obj, "Value");
        this.tag = ExtensionMethod.GetAttributeValue4String(obj, "Tag");
    };
    /**
     * @return {?}
     */
    SmsMessageVariable.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new SmsMessageVariable(this.Model);
    };
    /**
     * @return {?}
     */
    SmsMessageVariable.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var smsMessageVariable = new SmsMessageVariable(this.Model);
        smsMessageVariable.code = this.code;
        smsMessageVariable.value = this.value;
        smsMessageVariable.tag = this.tag;
        return smsMessageVariable;
    };
    SmsMessageVariable.Clr_Type_ID = "wf.SmsMessageVariable";
    return SmsMessageVariable;
}(ExtensionElement));
export { SmsMessageVariable };
if (false) {
    /** @type {?} */
    SmsMessageVariable.Clr_Type_ID;
    /** @type {?} */
    SmsMessageVariable.prototype.ClrTypeID;
    /** @type {?} */
    SmsMessageVariable.prototype.code;
    /** @type {?} */
    SmsMessageVariable.prototype.value;
    /** @type {?} */
    SmsMessageVariable.prototype.tag;
}
BpmnModelHelper.RegisterElementType(new SmsMessageVariable(new BpmnModel()));
