// @gsp-wf/wf-process-model/esm5/lib/model/TimeLimitControl.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionMethod, ExtensionElement } from "@edp-pmf/bpmn-model";
import { AutoHandleExpireOperation } from "./AutoHandleExpireOperation";
import { SendMessageExpireOperation } from "./SendMessageExpireOperation";
var TimeLimitControl = /** @class */ (function (_super) {
    tslib_1.__extends(TimeLimitControl, _super);
    function TimeLimitControl(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = TimeLimitControl.Clr_Type_ID;
        _this.expireOperations = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    TimeLimitControl.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    TimeLimitControl.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "Tag", this.tag);
        ExtensionMethod.SetAttributeValue(obj, "DueTime", this.dueTime);
        ExtensionMethod.SetAttributeValue(obj, "TimeUnit", this.timeUnit);
        ExtensionMethod.SetAttributeValue(obj, "Enable", this.enable);
        /** @type {?} */
        var expireOperationElements = new Array();
        if (this.expireOperations !== null && this.expireOperations.length > 0) {
            this.expireOperations.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                expireOperationElements.push(item.ConvertToJsonObject());
            }));
        }
        ExtensionMethod.SetAttributeValue(obj, "ExpireOperations", expireOperationElements);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    TimeLimitControl.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.tag = ExtensionMethod.GetAttributeValue4String(obj, "Tag");
        this.dueTime = ExtensionMethod.GetAttributeValue4String(obj, "DueTime");
        this.timeUnit = ExtensionMethod.GetAttributeValue4String(obj, "TimeUnit");
        this.enable = ExtensionMethod.GetAttributeValue4Boolean(obj, "Enable");
        /** @type {?} */
        var expireOperationElements = (/** @type {?} */ (ExtensionMethod.GetAttributeValue(obj, "ExpireOperations")));
        if (expireOperationElements !== null && expireOperationElements.length > 0) {
            expireOperationElements.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.OperationTypeName === 'SendMessageExpireOperation') {
                    /** @type {?} */
                    var expireOperation = new SendMessageExpireOperation(_this.Model);
                    expireOperation.LoadFromJsonObject(item);
                    _this.expireOperations.push(expireOperation);
                }
                else if (item.OperationTypeName === 'AutoHandleExpireOperation') {
                    /** @type {?} */
                    var expireOperation = new AutoHandleExpireOperation(_this.Model);
                    expireOperation.LoadFromJsonObject(item);
                    _this.expireOperations.push(expireOperation);
                }
            }));
        }
    };
    /**
     * @return {?}
     */
    TimeLimitControl.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new TimeLimitControl(this.Model);
    };
    /**
     * @return {?}
     */
    TimeLimitControl.prototype.clone = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        /** @type {?} */
        var timeLimitControl = new TimeLimitControl(this.Model);
        timeLimitControl.tag = this.tag;
        timeLimitControl.enable = this.enable;
        timeLimitControl.dueTime = this.dueTime;
        timeLimitControl.timeUnit = this.timeUnit;
        timeLimitControl.expireOperations = new Array();
        if (this.expireOperations && this.expireOperations.length > 0) {
            try {
                for (var _b = tslib_1.__values(this.expireOperations), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var e = _c.value;
                    if (e instanceof SendMessageExpireOperation) {
                        /** @type {?} */
                        var sendMessage = (/** @type {?} */ (e));
                        /** @type {?} */
                        var expireOperation = sendMessage.clone();
                        timeLimitControl.expireOperations.push(expireOperation);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        return timeLimitControl;
    };
    TimeLimitControl.Clr_Type_ID = "wf.TimeLimitControl";
    return TimeLimitControl;
}(ExtensionElement));
export { TimeLimitControl };
if (false) {
    /** @type {?} */
    TimeLimitControl.Clr_Type_ID;
    /** @type {?} */
    TimeLimitControl.prototype.ClrTypeID;
    /**
     * 自定义标签，用于区分期限控制方式
     * expression 表达式
     * constant 固定值
     * scope 时间范围
     * @type {?}
     */
    TimeLimitControl.prototype.tag;
    /**
     * 是否启用
     * @type {?}
     */
    TimeLimitControl.prototype.enable;
    /**
     * 限定时长
     * @type {?}
     */
    TimeLimitControl.prototype.dueTime;
    /**
     * 时间单位 day,hour,minute
     * @type {?}
     */
    TimeLimitControl.prototype.timeUnit;
    /** @type {?} */
    TimeLimitControl.prototype.expireOperations;
}
BpmnModelHelper.RegisterElementType(new TimeLimitControl(new BpmnModel()));
