// @gsp-wf/wf-process-model/esm5/lib/model/VariableSetting.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionMethod, ExtensionElement } from "@edp-pmf/bpmn-model";
var VariableSetting = /** @class */ (function (_super) {
    tslib_1.__extends(VariableSetting, _super);
    function VariableSetting(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = VariableSetting.Clr_Type_ID;
        return _this;
    }
    /**
     * @return {?}
     */
    VariableSetting.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        // ExtensionMethod.SetAttributeValue(obj, "VariableId", this.variableId);
        ExtensionMethod.SetAttributeValue(obj, "VariableCode", this.variableCode);
        ExtensionMethod.SetAttributeValue(obj, "VariableName", this.variableName);
        ExtensionMethod.SetAttributeValue(obj, "Tag", this.tag);
        ExtensionMethod.SetAttributeValue(obj, "ExecutionTime", this.executionTime);
        ExtensionMethod.SetAttributeValue(obj, "Value", this.value);
        ExtensionMethod.SetAttributeValue(obj, "ValueText", this.valueText);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    VariableSetting.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        // this.variableId = ExtensionMethod.GetAttributeValue4String(obj, "VariableId");
        this.variableCode = ExtensionMethod.GetAttributeValue4String(obj, "VariableCode");
        this.variableName = ExtensionMethod.GetAttributeValue4String(obj, "VariableName");
        this.tag = ExtensionMethod.GetAttributeValue4String(obj, "Tag");
        this.executionTime = ExtensionMethod.GetAttributeValue4String(obj, "ExecutionTime");
        this.value = ExtensionMethod.GetAttributeValue4String(obj, "Value");
        this.valueText = ExtensionMethod.GetAttributeValue4String(obj, "ValueText");
    };
    /**
     * @return {?}
     */
    VariableSetting.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new VariableSetting(this.Model);
    };
    /**
     * @return {?}
     */
    VariableSetting.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var variableSetting = new VariableSetting(this.Model);
        variableSetting.variableCode = this.variableCode;
        variableSetting.variableName = this.variableName;
        variableSetting.tag = this.tag;
        variableSetting.executionTime = this.executionTime;
        variableSetting.value = this.value;
        variableSetting.valueText = this.valueText;
        return variableSetting;
    };
    VariableSetting.Clr_Type_ID = "wf.VariableSetting";
    return VariableSetting;
}(ExtensionElement));
export { VariableSetting };
if (false) {
    /** @type {?} */
    VariableSetting.Clr_Type_ID;
    /** @type {?} */
    VariableSetting.prototype.ClrTypeID;
    /**
     * 变量编号
     * @type {?}
     */
    VariableSetting.prototype.variableCode;
    /**
     * 变量名称
     * @type {?}
     */
    VariableSetting.prototype.variableName;
    /**
     * 自定义标签，赋值方式
     * expression 语义化表达式
     * constant 常量
     * component 构件
     * @type {?}
     */
    VariableSetting.prototype.tag;
    /**
     * 变量执行时机
     * ACTIVITY_STARTED 活动开始
     * ACTIVITY_COMPLETED 活动完成
     * @type {?}
     */
    VariableSetting.prototype.executionTime;
    /**
     * 变量值
     * @type {?}
     */
    VariableSetting.prototype.value;
    /** @type {?} */
    VariableSetting.prototype.valueText;
}
BpmnModelHelper.RegisterElementType(new VariableSetting(new BpmnModel()));
