// @farris/devkit/esm5/lib/app/binding_data_manager.js
/*
 * @Author: Witt
 * @Date: 2018-12-29 10:46:01
 * @Last Modified by: Witt
 * @Last Modified time: 2018-12-30 17:56:02
 */
/**
 * BindingData管理类
 */
var BindingDataManager = /** @class */ (function () {
    /**
     * 构造函数
     */
    function BindingDataManager() {
        this.bindingDataMap = new Map();
    }
    /**
     * 获取BindingDataMap
     */
    BindingDataManager.prototype.getBindingDataMap = function () {
        return this.bindingDataMap;
    };
    /**
     * 根据name获取BindingData
     * @return 找不到时返回undefined
     */
    BindingDataManager.prototype.getBindingDataByName = function (name) {
        return this.bindingDataMap.get(name);
    };
    /**
     * 初始化全局的BindingData
     * @description 仅在高速模式时注册bindingData
     */
    BindingDataManager.prototype.regBindingData = function (name, bindingData) {
        this.bindingDataMap.set(name, bindingData);
    };
    BindingDataManager.prototype.unRegisteBindingData = function (name) {
        this.bindingDataMap.delete(name);
    };
    /**
     * 是否Repository已经存在
     */
    BindingDataManager.prototype.ifBindingDataExits = function (name) {
        var bindingData = this.getBindingDataByName(name);
        return bindingData ? true : false;
    };
    BindingDataManager.prototype.dispose = function () {
        this.bindingDataMap.clear();
    };
    return BindingDataManager;
}());
export { BindingDataManager };
