// @farris/devkit/esm5/lib/app/frame_component_refresher.js
/**
 * FrameComponent刷新器
 */
var FrameComponentRefresher = /** @class */ (function () {
    /**
     * 构造函数
     */
    function FrameComponentRefresher() {
        this.frameComponentMap = new Map();
    }
    /**
     * 刷新组件
     */
    FrameComponentRefresher.prototype.refreshComponents = function () {
        this.frameComponentMap.forEach(function (component, frameId) {
            if (typeof component['onFormLoad'] === 'function') {
                component.onFormLoad();
            }
        });
    };
    /**
     * 注册组件
     * @param frameId   FrameId
     * @param component Frame组件实例
     */
    FrameComponentRefresher.prototype.regFrameComponent = function (frameId, component) {
        this.frameComponentMap.set(frameId, component);
    };
    /**
     * 注销组件注册
     */
    FrameComponentRefresher.prototype.unregFrameContext = function (frameContext) {
        var frameId = frameContext.frameId;
        this.frameComponentMap.delete(frameId);
    };
    FrameComponentRefresher.prototype.dispose = function () {
        this.frameComponentMap.clear();
    };
    return FrameComponentRefresher;
}());
export { FrameComponentRefresher };
