// @farris/devkit/esm5/lib/app/render_context_metadata_manager.js
/**
 * ContextMetadata管理类
 */
var ContextMetadataManager = /** @class */ (function () {
    /**
     * 构造函数
     */
    function ContextMetadataManager() {
        this.contextMetadataMap = new Map();
    }
    /**
     * 获取ContextMetadataMap
     */
    ContextMetadataManager.prototype.getContextMetadataMap = function () {
        return this.contextMetadataMap;
    };
    /**
     * 根据name获取ContextMetadata
     * @return 找不到时返回undefined
     */
    ContextMetadataManager.prototype.getContextMetadataByName = function (name) {
        return this.contextMetadataMap.get(name);
    };
    /**
     * 初始化全局的ContextMetadata
     */
    ContextMetadataManager.prototype.register = function (name, contextMetadata) {
        this.contextMetadataMap.set(name, contextMetadata);
    };
    /**
     * ContextMetadata是否已经存在
     */
    ContextMetadataManager.prototype.exsit = function (name) {
        var contextMetadata = this.getContextMetadataByName(name);
        return contextMetadata ? true : false;
    };
    return ContextMetadataManager;
}());
export { ContextMetadataManager };
