// @farris/devkit/esm5/lib/app/repository_mananger.js
/*
 * @Author: Witt
 * @Date: 2018-12-29 10:46:01
 * @Last Modified by: Witt
 * @Last Modified time: 2018-12-30 18:06:11
 */
import { Injectable } from '@angular/core';
/**
 * Repository管理类
 */
var RepositoryManager = /** @class */ (function () {
    /**
     * 构造函数
     */
    function RepositoryManager() {
        this.repositoryMap = new Map();
    }
    /**
     * 注册Repository
     */
    RepositoryManager.prototype.regRepository = function (name, repository) {
        this.repositoryMap.set(name, repository);
    };
    RepositoryManager.prototype.unRegisteRepository = function (name) {
        this.repositoryMap.delete(name);
    };
    /**
     * 获取RepositoryMap
     * @internal
     */
    RepositoryManager.prototype.getRepositoryMap = function () {
        return this.repositoryMap;
    };
    /**
     * 获取Repository数组
     */
    RepositoryManager.prototype.getRepositories = function () {
        return Array.from(this.repositoryMap.values());
    };
    /**
     * 根据name获取Repository
     */
    RepositoryManager.prototype.getRepositoryByName = function (name) {
        return this.repositoryMap.get(name);
    };
    /**
     * 是否Repository已经存在
     */
    RepositoryManager.prototype.ifRepositoryExits = function (name) {
        var repository = this.getRepositoryByName(name);
        return repository ? true : false;
    };
    RepositoryManager.prototype.dispose = function () {
        // this.repositoryMap.forEach((repo: Repository<any>) => {
        //   if (repo) {
        //     repo.entityCollection.reset(false);
        //   }
        // });
        this.repositoryMap.clear();
    };
    RepositoryManager.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    RepositoryManager.ctorParameters = function () { return []; };
    return RepositoryManager;
}());
export { RepositoryManager };
