// @farris/devkit/esm5/lib/app/view_model_manager.js
/**
 * ViewModel管理类
 */
var ViewModelManager = /** @class */ (function () {
    /**
     * 构造函数
     */
    function ViewModelManager() {
        this.viewModelMap = new Map();
    }
    /**
     * 获取ViewModelMap
     */
    ViewModelManager.prototype.getViewModelMap = function () {
        return this.viewModelMap;
    };
    /**
     * 根据name获取ViewModel
     * @return 找不到时返回undefined
     */
    ViewModelManager.prototype.getViewModelByName = function (name) {
        return this.viewModelMap.get(name);
    };
    /**
     * 初始化全局的viewModel
     */
    ViewModelManager.prototype.register = function (name, viewModel) {
        this.viewModelMap.set(name, viewModel);
    };
    /**
     * 是否Repository已经存在
     */
    ViewModelManager.prototype.exsit = function (name) {
        var viewModel = this.getViewModelByName(name);
        return viewModel ? true : false;
    };
    ViewModelManager.prototype.dispose = function () {
        this.viewModelMap.clear();
    };
    return ViewModelManager;
}());
export { ViewModelManager };
