// @gsp-lcm/metadata-selector/fesm5/gsp-lcm-metadata-selector.js
import { CommonModule } from '@angular/common';
import { Observable as Observable$1 } from 'rxjs/Observable';
import { __values } from 'tslib';
import { FormsModule } from '@angular/forms';
import { TreeViewModule } from '@progress/kendo-angular-treeview';
import { GridModule } from '@progress/kendo-angular-grid';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { LayoutModule } from '@progress/kendo-angular-layout';
import { process } from '@progress/kendo-data-query';
import { DropDownsModule } from '@progress/kendo-angular-dropdowns';
import { LoadingService, LoadingModule } from '@farris/ui-loading';
import { distinctUntilChanged, debounceTime, throttleTime } from 'rxjs/operators';
import { Observable, Subject } from 'rxjs';
import { Injectable, Inject, EventEmitter, Component, Input, ViewChild, Output, Directive, HostListener, Renderer2, ElementRef, NgModule } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { GSP } from '@farris/ide-devkit';

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadatapackagedto.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MetadataPackageDto = /** @class */ (function () {
    function MetadataPackageDto(packageName, packageVersion, packageLocation, projectPath) {
        this.packageName = packageName;
        this.projectPath = projectPath;
        this.packageVersion = packageVersion;
        this.packageLocation = packageLocation;
    }
    return MetadataPackageDto;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var GSPMetadataService = /** @class */ (function () {
    function GSPMetadataService(http, url) {
        this.http = http;
        this.url = url;
        this.METADATA_BASE_PATH = '/api/dev/main/v1.0/metadatas';
        this.GSP_PROJECT_BASE_PATH = '/api/dev/main/v1.0/gsp-projects';
        this.Project_Base_Path = '/api/dev/main/v1.0/project';
        this.DBO_BASE_PATH = '/api/dev/main/v1.0/database-objects'; // dbo相关
        // dbo相关
        this.METADATA_PROJECT_BASE_PATH = '/api/dev/main/v1.0/metadata-projects';
        this.METADATA_PACKAGE_BASE_PATH = '/api/dev/main/v1.0/metadata-packages';
        this.METADATA_CONFIG_BASE_PATH = '/api/dev/main/v1.0/metadata-configs';
        // private Metadata4RefDto_Base_Path = '/api/dev/main/v1.0/metadata4refdtos'; pickMetadataIndex，dotnet不需要
        // tslint:disable-next-line:variable-name
        this.DevBasinBOInfo_Base_Path = '/api/dev/main/v1.0/dev-basic-boinfos'; // 业务对象
        // 业务对象
        this.NUGET_BASE_PATH = '/api/dev/main/v1.0/nuget-packages';
        this.SETTING_BASE_PATH = '/api/dev/main/v1.0/ide/settings';
        this.MAVEN_JAVA_BASE_PATH = '/api/dev/main/v1.0/repo-packages';
        this.FS_JAVA_BASE_PATH = '/api/dev/main/v1.0/file-system';
        this.JAVA_CODE_GENERATOR = '/api/dev/main/v1.0/generator';
        this.JAVA_BUILD = '/api/dev/main/v1.0/repo-packages';
        this.METADATA_JAVA_BASE_PATH = '/api/dev/main/v1.0/mdservice';
        this.SERVER_IP = (url === null || url === undefined) ? '' : url;
        this.sessionId = gsp.cache.get('sessionId');
    }
    /**
     * 元数据设计时服务
     */
    // 打包（功能触发）
    /**
     * 元数据设计时服务
     * @return {?}
     */
    // 打包（功能触发）
    GSPMetadataService.prototype.Pack = /**
     * 元数据设计时服务
     * @return {?}
     */
    // 打包（功能触发）
    function () {
        if (this.selectType === 'folder' || this.selectType === 'project') {
            if (this.selectPath !== '' && this.selectPath !== undefined) {
                this.PackageBuild(this.selectPath.substring(1));
            }
        }
        else {
            alert('请选择合适的目录');
        }
    };
    // 打包（外部调用）
    // 打包（外部调用）
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.PackageBuild = 
    // 打包（外部调用）
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var sessionId = gsp.cache.get('sessionId');
        /** @type {?} */
        var headers = new HttpHeaders({ 'Content-Type': 'application/json' }).append('SessionId', sessionId);
        /** @type {?} */
        var packageDto = new MetadataPackageDto('', '', '', path);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PACKAGE_BASE_PATH;
        return this.http.post(url, packageDto, {
            headers: headers,
            observe: 'response'
        });
    };
    // 递归查找是否存在工程文件
    // 递归查找是否存在工程文件
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.IsProjExistRecursively = 
    // 递归查找是否存在工程文件
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/validationrecursively?path=' + path;
        /** @type {?} */
        var metadata = new Subject();
        this.http.get(url).subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            // if (result) {
            metadata.next(result);
            // }
        }));
        return metadata;
    };
    // 初始化元数据实体
    // 初始化元数据实体
    /**
     * @param {?} metadataDto
     * @return {?}
     */
    GSPMetadataService.prototype.InitializeMetadataEntity = 
    // 初始化元数据实体
    /**
     * @param {?} metadataDto
     * @return {?}
     */
    function (metadataDto) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        // tslint:disable-next-line:max-line-length
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_BASE_PATH + '/initialized?nameSpace=' + metadataDto.nameSpace + '&code=' + metadataDto.code + '&name=' + metadataDto.name + '&type=' + metadataDto.type + '&bizObjectID=' + metadataDto.bizobjectID + '&metadataPath=' + metadataDto.relativePath + '&extendProperty=' + metadataDto.extendProperty;
        return this.http.get(url, { headers: headers });
    };
    // 新建元数据
    // 新建元数据
    /**
     * @param {?} metadataDto
     * @return {?}
     */
    GSPMetadataService.prototype.CreateMetadata = 
    // 新建元数据
    /**
     * @param {?} metadataDto
     * @return {?}
     */
    function (metadataDto) {
        /** @type {?} */
        var content = {
            'ID': metadataDto.id,
            'NameSpace': metadataDto.nameSpace,
            'Code': metadataDto.code,
            'Name': metadataDto.name,
            'FileName': metadataDto.fileName,
            'RelativePath': metadataDto.relativePath,
            'Content': metadataDto.content,
            'Type': metadataDto.type,
            'BizobjectID': metadataDto.bizobjectID,
            'ExtendProperty': metadataDto.extendProperty
        };
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_BASE_PATH;
        return this.http.post(url, content, {
            params: {},
            headers: headers,
            observe: 'response'
        });
    };
    // 删除元数据文件
    // 删除元数据文件
    /**
     * @return {?}
     */
    GSPMetadataService.prototype.DeleteMetadata = 
    // 删除元数据文件
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.selectPath !== '' && this.selectPath !== undefined) {
            if (this.selectType === 'file') {
                this.DeleteMetadataOrDir(this.selectPath.substring(1)).subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                function (result) {
                    // 刷新节点
                    // this.treeModel.removeFileNode();
                    alert('删除成功！');
                    _this.selectPath = '';
                    _this.selectType = '';
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    alert('删除失败！');
                }));
            }
            else {
                alert('请选择正确的文件进行删除！');
            }
        }
        else {
            alert('请选择正确的文件进行删除！');
        }
    };
    // 保存元数据
    // 保存元数据
    /**
     * @param {?} metadataDto
     * @param {?} sessionId
     * @return {?}
     */
    GSPMetadataService.prototype.SaveMetadata = 
    // 保存元数据
    /**
     * @param {?} metadataDto
     * @param {?} sessionId
     * @return {?}
     */
    function (metadataDto, sessionId) {
        /** @type {?} */
        var content = {
            'ID': metadataDto.id,
            'NameSpace': metadataDto.nameSpace,
            'Code': metadataDto.code,
            'Name': metadataDto.name,
            'FileName': metadataDto.fileName,
            'RelativePath': metadataDto.relativePath,
            'Content': metadataDto.content,
            'Type': metadataDto.type,
            'BizobjectID': metadataDto.bizobjectID,
            'ExtendProperty': metadataDto.extendProperty,
            'Extendable': metadataDto.extendable
        };
        // tslint:disable-next-line:one-variable-per-declaration
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_BASE_PATH;
        return this.http.put(url, content, {
            headers: headers,
            observe: 'response'
        });
    };
    // Load元数据（外部调用）
    // Load元数据（外部调用）
    /**
     * @param {?} fullName
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.LoadMetadata = 
    // Load元数据（外部调用）
    /**
     * @param {?} fullName
     * @param {?} path
     * @return {?}
     */
    function (fullName, path) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var metadataFullPath = path.replace(/\\/g, '/') + '/' + fullName;
        /** @type {?} */
        var encMetadataFullPath = encodeURIComponent(metadataFullPath);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_BASE_PATH + '/load?metadataFullPath=' + encMetadataFullPath;
        return this.http.get(url, { headers: headers });
    };
    // Load bo内元数据（外部调用）
    // Load bo内元数据（外部调用）
    /**
     * @param {?} fullName
     * @param {?} path
     * @param {?} currentProjPath
     * @return {?}
     */
    GSPMetadataService.prototype.LoadMetadataInBo = 
    // Load bo内元数据（外部调用）
    /**
     * @param {?} fullName
     * @param {?} path
     * @param {?} currentProjPath
     * @return {?}
     */
    function (fullName, path, currentProjPath) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var metadataFullPath = path.replace(/\\/g, '/') + '/' + fullName;
        /** @type {?} */
        var encMetadataFullPath = encodeURIComponent(metadataFullPath);
        // tslint:disable-next-line: max-line-length
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_BASE_PATH + '/metadataInBo?metadataFullPath=' + encMetadataFullPath
            + '&currentProjPath=' + currentProjPath;
        return this.http.get(url, { headers: headers });
    };
    // 删除文件夹
    // 删除文件夹
    /**
     * @return {?}
     */
    GSPMetadataService.prototype.DeleteFolder = 
    // 删除文件夹
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.selectPath !== '' && this.selectPath !== undefined) {
            if (this.selectType === 'folder') {
                this.DeleteMetadataOrDir(this.selectPath.substring(1)).subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                function (result) {
                    // 刷新节点
                    // this.treeModel.removeFolderNode();
                    alert('删除成功！');
                    _this.selectPath = '';
                    _this.selectType = '';
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    alert('删除失败！');
                }));
            }
            else {
                alert('请选择正确的目录进行删除！');
            }
        }
        else {
            alert('请选择正确的目录进行删除！');
        }
    };
    // 删除工程
    // 删除工程
    /**
     * @return {?}
     */
    GSPMetadataService.prototype.DeleteProj = 
    // 删除工程
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.selectPath !== '' && this.selectPath !== undefined) {
            if (this.selectType === 'project') {
                this.DeleteProject(this.selectPath.substring(1)).subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                function (result) {
                    // 刷新节点
                    // this.treeModel.removeProjectNode();
                    alert('删除成功！');
                    _this.selectPath = '';
                    _this.selectType = '';
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    alert('删除失败！');
                }));
            }
            else {
                alert('请选择正确的工程节点进行删除！');
            }
        }
        else {
            alert('请选择正确的工程节点进行删除！');
        }
    };
    // 元数据或者文件夹的删除（外部调用）
    // 元数据或者文件夹的删除（外部调用）
    /**
     * @param {?} fullPath
     * @return {?}
     */
    GSPMetadataService.prototype.DeleteMetadataOrDir = 
    // 元数据或者文件夹的删除（外部调用）
    /**
     * @param {?} fullPath
     * @return {?}
     */
    function (fullPath) {
        /** @type {?} */
        var sessionId = gsp.cache.get('sessionId');
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', sessionId);
        /** @type {?} */
        var encFullPath = encodeURIComponent(fullPath.replace(/\\/g, '/'));
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_BASE_PATH + '/delete?fullPath=' + encFullPath;
        return this.http.delete(url, {
            params: {},
            headers: headers,
            observe: 'response'
        });
    };
    // 判断元数据是否被依赖
    // 判断元数据是否被依赖
    /**
     * @param {?} fullPath
     * @return {?}
     */
    GSPMetadataService.prototype.isMetadataRefed = 
    // 判断元数据是否被依赖
    /**
     * @param {?} fullPath
     * @return {?}
     */
    function (fullPath) {
        /** @type {?} */
        var sessionId = gsp.cache.get('sessionId');
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', sessionId);
        /** @type {?} */
        var encFullPath = encodeURIComponent(fullPath.replace(/\\/g, '/'));
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/isMetadataRefed?path=' + encFullPath;
        return this.http.get(url, {
            params: {},
            headers: headers,
            observe: 'response'
        });
    };
    // 删除工程
    // 删除工程
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.DeleteProject = 
    // 删除工程
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var sessionId = gsp.cache.get('sessionId');
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.Project_Base_Path + '/' + path.replace(/\\/g, '/');
        return this.http.delete(url, {
            params: {},
            headers: headers,
            observe: 'response'
        });
    };
    /**
     * @return {?}
     */
    GSPMetadataService.prototype.GetDevRootPath = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.Project_Base_Path + '/rootpath';
        /** @type {?} */
        var metadata = new Subject();
        this.http.get(url, { responseType: 'text' }).subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            if (result) {
                metadata.next(result);
            }
        }));
        return metadata;
    };
    /**
     * @param {?} fullName
     * @param {?} path
     * @param {?} sessionId
     * @return {?}
     */
    GSPMetadataService.prototype.LoadDBO = /**
     * @param {?} fullName
     * @param {?} path
     * @param {?} sessionId
     * @return {?}
     */
    function (fullName, path, sessionId) {
        /** @type {?} */
        var dboFullPath = path.replace(/\\/g, '/') + '/' + fullName;
        /** @type {?} */
        var url = this.SERVER_IP + this.DBO_BASE_PATH + '/' + dboFullPath;
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        return this.http.get(url, { headers: headers });
    };
    // 新建文件夹
    // 新建文件夹
    /**
     * @param {?} path
     * @param {?} name
     * @return {?}
     */
    GSPMetadataService.prototype.CreateDir = 
    // 新建文件夹
    /**
     * @param {?} path
     * @param {?} name
     * @return {?}
     */
    function (path, name) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.Project_Base_Path + '/folder' + path + '/' + name + '?quiet=true';
        return this.http.post(url, null, {
            params: {},
            headers: headers,
            observe: 'response'
        });
    };
    // 获取元数据工程信息
    // 获取元数据工程信息
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.GetProjectInfo = 
    // 获取元数据工程信息
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '?path=' + path;
        return this.http.get(url, { headers: headers });
    };
    // 获取工程信息
    // 获取工程信息
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.GetGspProjInfo = 
    // 获取工程信息
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.GSP_PROJECT_BASE_PATH + '?path=' + path;
        return this.http.get(url, { headers: headers });
    };
    // 获取工程路径
    // 获取工程路径
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.GetProjPath = 
    // 获取工程路径
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/path?path=' + path;
        return this.http.get(url, { responseType: 'text' });
    };
    // 获取当前工程下或者其他元数据包中依赖的元数据
    // 获取当前工程下或者其他元数据包中依赖的元数据
    /**
     * @param {?} path
     * @param {?} metadataid
     * @param {?} sessionId
     * @return {?}
     */
    GSPMetadataService.prototype.GetRefMetadata = 
    // 获取当前工程下或者其他元数据包中依赖的元数据
    /**
     * @param {?} path
     * @param {?} metadataid
     * @param {?} sessionId
     * @return {?}
     */
    function (path, metadataid, sessionId) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_BASE_PATH + '/relied?metadataPath=' + path + '&metadataID=' + metadataid;
        return this.http.get(url, { headers: headers });
    };
    // 获取当前工程下所有的元数据包
    // 获取当前工程下所有的元数据包
    /**
     * @param {?} spacePath
     * @param {?} typeName
     * @return {?}
     */
    GSPMetadataService.prototype.GetMetadataList = 
    // 获取当前工程下所有的元数据包
    /**
     * @param {?} spacePath
     * @param {?} typeName
     * @return {?}
     */
    function (spacePath, typeName) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '?path=' + spacePath + '&metadataTypeList=' + typeName;
        return this.http.get(url, { headers: headers });
    };
    // 获取bo工程下所有的元数据列表
    // 获取bo工程下所有的元数据列表
    /**
     * @param {?} projPath
     * @param {?} typeName
     * @return {?}
     */
    GSPMetadataService.prototype.GetMetadataListInBo = 
    // 获取bo工程下所有的元数据列表
    /**
     * @param {?} projPath
     * @param {?} typeName
     * @return {?}
     */
    function (projPath, typeName) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/metadataListInBo?path=' + projPath + '&metadataTypeList=' + typeName;
        return this.http.get(url, { headers: headers });
    };
    /**
     * @param {?} path
     * @param {?} typeCode
     * @param {?} filter
     * @param {?} postfix
     * @return {?}
     */
    GSPMetadataService.prototype.GetMetadataByFilter = /**
     * @param {?} path
     * @param {?} typeCode
     * @param {?} filter
     * @param {?} postfix
     * @return {?}
     */
    function (path, typeCode, filter, postfix) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_BASE_PATH + '/filter?path=' + path +
            '&typeCode=' + typeCode + '&filter=' + filter + '&postfix=' + postfix;
        return this.http.get(url, { headers: headers });
    };
    // pickMetadataIndex(scope: number, currentProjPath: string, metadataIndexDto: any): Observable<any> {
    //   const sessionId = gsp.cache.get('sessionId');
    //   const headers = new HttpHeaders({ 'Content-Type': 'application/json' }).append('SessionId', sessionId);
    //   const url = this.SERVER_IP + this.Metadata4RefDto_Base_Path + '/scope/' + scope + '?currentPath=' + currentProjPath;
    //   return this.http.post(url, metadataIndexDto, {
    //     headers
    //   });
    // }
    // pickMetadataIndex(scope: number, currentProjPath: string, metadataIndexDto: any): Observable<any> {
    //   const sessionId = gsp.cache.get('sessionId');
    //   const headers = new HttpHeaders({ 'Content-Type': 'application/json' }).append('SessionId', sessionId);
    //   const url = this.SERVER_IP + this.Metadata4RefDto_Base_Path + '/scope/' + scope + '?currentPath=' + currentProjPath;
    //   return this.http.post(url, metadataIndexDto, {
    //     headers
    //   });
    // }
    /**
     * @param {?} scope
     * @param {?} currentProjPath
     * @param {?} metadataIndexDto
     * @return {?}
     */
    GSPMetadataService.prototype.pickMetadataIndex = 
    // pickMetadataIndex(scope: number, currentProjPath: string, metadataIndexDto: any): Observable<any> {
    //   const sessionId = gsp.cache.get('sessionId');
    //   const headers = new HttpHeaders({ 'Content-Type': 'application/json' }).append('SessionId', sessionId);
    //   const url = this.SERVER_IP + this.Metadata4RefDto_Base_Path + '/scope/' + scope + '?currentPath=' + currentProjPath;
    //   return this.http.post(url, metadataIndexDto, {
    //     headers
    //   });
    // }
    /**
     * @param {?} scope
     * @param {?} currentProjPath
     * @param {?} metadataIndexDto
     * @return {?}
     */
    function (scope, currentProjPath, metadataIndexDto) {
        /** @type {?} */
        var sessionId = gsp.cache.get('sessionId');
        /** @type {?} */
        var headers = new HttpHeaders({ 'Content-Type': 'application/json' }).append('SessionId', sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/pickmd?scopeType=' + scope + '&currentPath=' + currentProjPath;
        return this.http.post(url, metadataIndexDto, {
            headers: headers
        });
    };
    // 获取元数据包中所包含的元数据信息
    // 获取元数据包中所包含的元数据信息
    /**
     * @param {?} packagePath
     * @param {?} packageName
     * @return {?}
     */
    GSPMetadataService.prototype.GetMetadataPackageInfo = 
    // 获取元数据包中所包含的元数据信息
    /**
     * @param {?} packagePath
     * @param {?} packageName
     * @return {?}
     */
    function (packagePath, packageName) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var packageFullName = '';
        if (packagePath) {
            packageFullName = packagePath.replace(/\\/g, '/') + '/' + packageName;
        }
        else {
            packageFullName = packageName;
        }
        /** @type {?} */
        var encPackageFullName = encodeURIComponent(packageFullName);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PACKAGE_BASE_PATH + '/packagePath?packagePath=' + encPackageFullName;
        return this.http.get(url, { headers: headers });
    };
    // 获取本地包
    // 获取本地包
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.GetLocalPackages = 
    // 获取本地包
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PACKAGE_BASE_PATH + '/local?path=' + path;
        return this.http.get(url, { headers: headers });
    };
    /**
     * @param {?} scope
     * @param {?=} metadataType
     * @param {?=} searchText
     * @param {?=} page
     * @param {?=} selectedPackageSource
     * @param {?=} currentProjPath
     * @return {?}
     */
    GSPMetadataService.prototype.GetMetadataInfosByScope = /**
     * @param {?} scope
     * @param {?=} metadataType
     * @param {?=} searchText
     * @param {?=} page
     * @param {?=} selectedPackageSource
     * @param {?=} currentProjPath
     * @return {?}
     */
    function (scope, metadataType, searchText, page, selectedPackageSource, currentProjPath) {
        if (metadataType === void 0) { metadataType = null; }
        if (searchText === void 0) { searchText = null; }
        if (page === void 0) { page = null; }
        if (selectedPackageSource === void 0) { selectedPackageSource = null; }
        if (currentProjPath === void 0) { currentProjPath = ''; }
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_BASE_PATH + '/scope/' + scope + '?path=' + currentProjPath;
        if (page) {
            url = url + '&page=true&pageIndex=' + page.pageIndex + '&pageSize=' + page.pageSize;
        }
        else {
            url = url + '&page=false';
        }
        if (metadataType) {
            url = url + '&metadataTypeList=' + metadataType;
        }
        if (searchText) {
            url = url + '&text=' + searchText;
        }
        if ((scope === 2 || scope === 3) && selectedPackageSource) {
            url = url + '&selectedPackageSource=' + selectedPackageSource;
        }
        return this.http.get(url, { headers: headers });
    };
    /**
     * @param {?} scope
     * @param {?=} metadataType
     * @param {?=} searchText
     * @param {?=} page
     * @param {?=} selectedPackageSource
     * @param {?=} currentProjPath
     * @return {?}
     */
    GSPMetadataService.prototype.GetMetadataInfosByScope4Java = /**
     * @param {?} scope
     * @param {?=} metadataType
     * @param {?=} searchText
     * @param {?=} page
     * @param {?=} selectedPackageSource
     * @param {?=} currentProjPath
     * @return {?}
     */
    function (scope, metadataType, searchText, page, selectedPackageSource, currentProjPath) {
        if (metadataType === void 0) { metadataType = null; }
        if (searchText === void 0) { searchText = null; }
        if (page === void 0) { page = null; }
        if (selectedPackageSource === void 0) { selectedPackageSource = null; }
        if (currentProjPath === void 0) { currentProjPath = ''; }
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/mdlist?scopeType=' + scope + '&path=' + currentProjPath;
        if (page) {
            url = url + '&page=true&pageIndex=' + page.pageIndex + '&pageSize=' + page.pageSize;
        }
        else {
            url = url + '&page=false';
        }
        if (metadataType) {
            url = url + '&metadataTypeList=' + metadataType;
        }
        if (searchText) {
            url = url + '&text=' + searchText;
        }
        if ((scope === 2 || scope === 3) && selectedPackageSource) {
            url = url + '&selectedPackageSource=' + selectedPackageSource;
        }
        return this.http.get(url, { headers: headers });
    };
    // 获取元数据以及包的信息
    // GetMetadataFromPackage(packageName: string, packagePath: string, metadataID: string): Observable<any> {
    //   const headers = new HttpHeaders().set('SessionId', this.sessionId);
    //   let fullName = '';
    //   if (packageName.endsWith('mdpkg')) {
    //     fullName = packageName;
    //   } else {
    //     fullName = packageName + '.mdpkg';
    //   }
    //   let path = '';
    //   if (packagePath) {
    //     path = packagePath;
    //   }
    //   const url = this.SERVER_IP + this.Metadata4RefDto_Base_Path + '?packageName=' + fullName +
    //     '&packagePath=' + path + '&metadataID=' + metadataID;
    //   return this.http.get(url, { headers });
    // }
    // 获取某个路径下的所有元数据包
    // 获取元数据以及包的信息
    // GetMetadataFromPackage(packageName: string, packagePath: string, metadataID: string): Observable<any> {
    //   const headers = new HttpHeaders().set('SessionId', this.sessionId);
    //   let fullName = '';
    //   if (packageName.endsWith('mdpkg')) {
    //     fullName = packageName;
    //   } else {
    //     fullName = packageName + '.mdpkg';
    //   }
    //   let path = '';
    //   if (packagePath) {
    //     path = packagePath;
    //   }
    //   const url = this.SERVER_IP + this.Metadata4RefDto_Base_Path + '?packageName=' + fullName +
    //     '&packagePath=' + path + '&metadataID=' + metadataID;
    //   return this.http.get(url, { headers });
    // }
    // 获取某个路径下的所有元数据包
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.GetMetadataPackages = 
    // 获取元数据以及包的信息
    // GetMetadataFromPackage(packageName: string, packagePath: string, metadataID: string): Observable<any> {
    //   const headers = new HttpHeaders().set('SessionId', this.sessionId);
    //   let fullName = '';
    //   if (packageName.endsWith('mdpkg')) {
    //     fullName = packageName;
    //   } else {
    //     fullName = packageName + '.mdpkg';
    //   }
    //   let path = '';
    //   if (packagePath) {
    //     path = packagePath;
    //   }
    //   const url = this.SERVER_IP + this.Metadata4RefDto_Base_Path + '?packageName=' + fullName +
    //     '&packagePath=' + path + '&metadataID=' + metadataID;
    //   return this.http.get(url, { headers });
    // }
    // 获取某个路径下的所有元数据包
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.Project_Base_Path + '/children/' + path;
        return this.http.get(url, { headers: headers });
    };
    // 获取元数据类型及后缀信息
    // 获取元数据类型及后缀信息
    /**
     * @return {?}
     */
    GSPMetadataService.prototype.GetMetadataTypeList = 
    // 获取元数据类型及后缀信息
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_CONFIG_BASE_PATH;
        return this.http.get(url, { headers: headers });
    };
    // 获取后缀
    // 获取后缀
    /**
     * @param {?} metadataType
     * @return {?}
     */
    GSPMetadataService.prototype.GetPostfix = 
    // 获取后缀
    /**
     * @param {?} metadataType
     * @return {?}
     */
    function (metadataType) {
        this.GetMetadataTypeList().subscribe((/**
         * @param {?} typeList
         * @return {?}
         */
        function (typeList) {
            typeList.forEach((/**
             * @param {?} type
             * @return {?}
             */
            function (type) {
                if (metadataType.toLowerCase() === type.typeCode.toLowerCase()) {
                    return type.postfix;
                }
            }));
        }));
        return new Observable();
    };
    // 更新工程引用
    // 更新工程引用
    /**
     * @param {?} projectPath
     * @param {?} packageHeader
     * @return {?}
     */
    GSPMetadataService.prototype.UpdateMetadataProject = 
    // 更新工程引用
    /**
     * @param {?} projectPath
     * @param {?} packageHeader
     * @return {?}
     */
    function (projectPath, packageHeader) {
        /** @type {?} */
        var content = {
            ProjectPath: projectPath,
            DepPackageName: packageHeader.name,
            DepPackageVersion: packageHeader.version.versionString,
            DepPackageLocation: packageHeader.location
        };
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH;
        return this.http.put(url, content, {
            params: {},
            headers: headers,
            observe: 'response'
        });
    };
    // 根据业务对象ID获取app、su等信息
    // 根据业务对象ID获取app、su等信息
    /**
     * @param {?} businessObjectID
     * @param {?} sessionId
     * @return {?}
     */
    GSPMetadataService.prototype.GetDevBasicBOInfo = 
    // 根据业务对象ID获取app、su等信息
    /**
     * @param {?} businessObjectID
     * @param {?} sessionId
     * @return {?}
     */
    function (businessObjectID, sessionId) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.DevBasinBOInfo_Base_Path + '/' + businessObjectID;
        return this.http.get(url, { headers: headers });
    };
    /**
     * @param {?} path
     * @param {?} fileName
     * @return {?}
     */
    GSPMetadataService.prototype.ValidateRepeatName = /**
     * @param {?} path
     * @param {?} fileName
     * @return {?}
     */
    function (path, fileName) {
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_BASE_PATH + '/validation?path=' + path + '&fileName=' + fileName;
        /** @type {?} */
        var metadata = new Subject();
        this.http.get(url).subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            // if (result) {
            metadata.next(result);
            // }
        }));
        return metadata;
    };
    /**
     * @param {?} path
     * @param {?} sessionId
     * @return {?}
     */
    GSPMetadataService.prototype.SetMetadataUri = /**
     * @param {?} path
     * @param {?} sessionId
     * @return {?}
     */
    function (path, sessionId) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_CONFIG_BASE_PATH + '?path=' + path;
        return this.http.put(url, null, {
            params: {},
            headers: headers,
            observe: 'response'
        });
    };
    /**
     * @return {?}
     */
    GSPMetadataService.prototype.getGspProjectType = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.GSP_PROJECT_BASE_PATH + '/type';
        return this.http.get(url, { headers: headers });
    };
    /**
     * @param {?} path
     * @param {?} codeLanguage
     * @return {?}
     */
    GSPMetadataService.prototype.updateGspProject = /**
     * @param {?} path
     * @param {?} codeLanguage
     * @return {?}
     */
    function (path, codeLanguage) {
        /** @type {?} */
        var sessionId = gsp.cache.get('sessionId');
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.GSP_PROJECT_BASE_PATH + '?path=' + path + '&codeLanguage=' + codeLanguage;
        return this.http.put(url, null, {
            params: {},
            headers: headers,
            observe: 'response'
        });
    };
    /**
     * @return {?}
     */
    GSPMetadataService.prototype.getNugetSourceList = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.NUGET_BASE_PATH + '/sources';
        return this.http.get(url, { headers: headers });
    };
    /**
     * @return {?}
     */
    GSPMetadataService.prototype.getMavenSourceList = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.MAVEN_JAVA_BASE_PATH + '/repolist';
        return this.http.get(url, { headers: headers });
    };
    // 获取技术栈
    // tslint:disable-next-line:adjacent-overload-signatures
    // 获取技术栈
    // tslint:disable-next-line:adjacent-overload-signatures
    /**
     * @param {?} projPath
     * @return {?}
     */
    GSPMetadataService.prototype.getCodeLanguage = 
    // 获取技术栈
    // tslint:disable-next-line:adjacent-overload-signatures
    /**
     * @param {?} projPath
     * @return {?}
     */
    function (projPath) {
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/codelanguage?projPath=' + projPath;
        return this.http.get(url, { responseType: 'text' });
    };
    // 获取setting信息
    // 获取setting信息
    /**
     * @return {?}
     */
    GSPMetadataService.prototype.getSettingInfo = 
    // 获取setting信息
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.SETTING_BASE_PATH;
        return this.http.get(url, { headers: headers });
    };
    // 是否开启maven
    // 是否开启maven
    /**
     * @return {?}
     */
    GSPMetadataService.prototype.getMavenSetting = 
    // 是否开启maven
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_BASE_PATH + '/mavensetting';
        return this.http.get(url);
    };
    // 获取bo路径
    // 获取bo路径
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.getBoPath = 
    // 获取bo路径
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/bopath?path=' + path;
        return this.http.get(url, { responseType: 'text' });
    };
    // 获取是否是maven引用
    // 获取是否是maven引用
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.getMavenStatus = 
    // 获取是否是maven引用
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/isMaven?path=' + path;
        return this.http.get(url);
    };
    // 获取是否是解析型工具
    // 获取是否是解析型工具
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.isInterpretation = 
    // 获取是否是解析型工具
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/isInterpretation?path=' + path;
        return this.http.get(url);
    };
    // 生成代码
    // 生成代码
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.generate = 
    // 生成代码
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var url = this.JAVA_CODE_GENERATOR + '?path=' + path;
        return this.http.post(url, null);
    };
    // 获取ProcessMode
    // 获取ProcessMode
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.getProcessMode = 
    // 获取ProcessMode
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/processMode?path=' + path;
        return this.http.get(url, { responseType: 'text' });
    };
    // 添加依赖并更新版本
    // 添加依赖并更新版本
    /**
     * @param {?} path
     * @param {?} metadataPackages
     * @param {?} isMaven
     * @return {?}
     */
    GSPMetadataService.prototype.addDependencyAndRestore = 
    // 添加依赖并更新版本
    /**
     * @param {?} path
     * @param {?} metadataPackages
     * @param {?} isMaven
     * @return {?}
     */
    function (path, metadataPackages, isMaven) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url;
        if (isMaven) {
            url = this.SERVER_IP + this.MAVEN_JAVA_BASE_PATH + '/adddepedency?path=' + path
                + '&metadataPackages=' + metadataPackages;
            return this.http.put(url, null, {
                params: {},
                headers: headers,
                observe: 'response'
            });
        }
        else {
            url = this.SERVER_IP + this.NUGET_BASE_PATH + '/adddepedency?projPath=' + path
                + '&metadataPackages=' + metadataPackages;
            return this.http.post(url, null, {
                params: {},
                headers: headers,
                observe: 'response'
            });
        }
    };
    // 在资源管理器中打开
    // 在资源管理器中打开
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.showInExplorer = 
    // 在资源管理器中打开
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var url = this.SERVER_IP + this.FS_JAVA_BASE_PATH + '?path=' + path;
        return this.http.get(url);
    };
    /**
     * @param {?} path
     * @param {?} metadataPackages
     * @return {?}
     */
    GSPMetadataService.prototype.restoreMavenPackage = /**
     * @param {?} path
     * @param {?} metadataPackages
     * @return {?}
     */
    function (path, metadataPackages) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.MAVEN_JAVA_BASE_PATH + '/adddepedency?path=' + path + '&metadataPackages=' + metadataPackages;
        return this.http.put(url, null, {
            headers: headers,
            observe: 'response'
        });
    };
    // 获取菜单
    // 获取菜单
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.getMenu = 
    // 获取菜单
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var url = this.SERVER_IP + this.SETTING_BASE_PATH + '/menu?path=' + path;
        return this.http.get(url);
    };
    /**
     * @param {?} projPath
     * @param {?=} fakeFlag
     * @return {?}
     */
    GSPMetadataService.prototype.mavenBuildAll = /**
     * @param {?} projPath
     * @param {?=} fakeFlag
     * @return {?}
     */
    function (projPath, fakeFlag) {
        if (fakeFlag === void 0) { fakeFlag = false; }
        if (fakeFlag) {
            return this.getMenu(projPath);
        }
        /** @type {?} */
        var url = this.JAVA_BUILD + '/compileAll?path=' + projPath;
        return this.http.post(url, null);
    };
    // 获取路径下的工程路径
    // 获取路径下的工程路径
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.getProjPathsInPath = 
    // 获取路径下的工程路径
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/projinpath?path=' + path;
        return this.http.get(url);
    };
    /**
     * @param {?} path
     * @param {?} mdpkgName
     * @return {?}
     */
    GSPMetadataService.prototype.getMdpkgNameExistInBo = /**
     * @param {?} path
     * @param {?} mdpkgName
     * @return {?}
     */
    function (path, mdpkgName) {
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/mdpkgNameExist?path=' + path + '&mdpkgName=' + mdpkgName;
        return this.http.get(url, { responseType: 'text' });
    };
    // 是否存在工程文件，废弃，使用getProjPath代替
    // 是否存在工程文件，废弃，使用getProjPath代替
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.IsProjExist = 
    // 是否存在工程文件，废弃，使用getProjPath代替
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/validation?path=' + path;
        /** @type {?} */
        var metadata = new Subject();
        this.http.get(url).subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            metadata.next(result);
        }));
        return metadata;
    };
    // 获取Nuget包下所有的元数据包，nuget已废弃
    // 获取Nuget包下所有的元数据包，nuget已废弃
    /**
     * @param {?} path
     * @return {?}
     */
    GSPMetadataService.prototype.GetMetadataPackagesUnderNugetDir = 
    // 获取Nuget包下所有的元数据包，nuget已废弃
    /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        var url = this.SERVER_IP + this.METADATA_PACKAGE_BASE_PATH + '?path=' + path;
        return this.http.get(url, { headers: headers });
    };
    GSPMetadataService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    GSPMetadataService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: String, decorators: [{ type: Inject, args: ['url',] }] }
    ]; };
    return GSPMetadataService;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/metadata.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// mock数据
/** @type {?} */
var classifier = [
    { id: 'project', text: '工程内元数据' },
    { id: 'local', text: '离线的元数据' },
    { id: 'remote', text: '仓库中元数据' },
    { id: 'ref', text: '引用的元数据' } // 3
];
/** @type {?} */
var classifierForBo = [
    { id: 'bo', text: 'BO内元数据' },
    { id: 'local', text: '离线的元数据' },
    { id: 'remote', text: '仓库中元数据' },
    { id: 'ref', text: '引用的元数据' } // 3
];
var MetadataService = /** @class */ (function () {
    function MetadataService(gspMetadataService) {
        this.gspMetadataService = gspMetadataService;
        this.selectedChange = new EventEmitter();
        // tslint:disable-next-line:member-ordering
        this.metadatas = [];
        // tslint:disable-next-line:member-ordering
        this.packages = [];
        this.packRefPath = '/packages';
    }
    /**
     * @param {?} type
     * @return {?}
     */
    MetadataService.prototype.getMetadataClassifier = /**
     * @param {?} type
     * @return {?}
     */
    function (type) {
        switch (type) {
            case 'bo':
                return classifierForBo;
            case 'project':
            default:
                return classifier;
        }
    };
    /**
     * @param {?} metadata
     * @return {?}
     */
    MetadataService.prototype.addMetadata = /**
     * @param {?} metadata
     * @return {?}
     */
    function (metadata) {
        this.metadatas.push(metadata);
    };
    /**
     * @return {?}
     */
    MetadataService.prototype.clearMetadata = /**
     * @return {?}
     */
    function () {
        this.metadatas = [];
    };
    /**
     * @param {?} mp
     * @return {?}
     */
    MetadataService.prototype.addPackage = /**
     * @param {?} mp
     * @return {?}
     */
    function (mp) {
        this.packages.push(mp);
    };
    /**
     * @return {?}
     */
    MetadataService.prototype.clearPackage = /**
     * @return {?}
     */
    function () {
        this.packages = [];
    };
    // 获取当前工程下所有的元数据
    // 获取当前工程下所有的元数据
    /**
     * @param {?} typeName
     * @return {?}
     */
    MetadataService.prototype.GetMetadataList = 
    // 获取当前工程下所有的元数据
    /**
     * @param {?} typeName
     * @return {?}
     */
    function (typeName) {
        return this.gspMetadataService.GetMetadataList(this.currentProjectPath, typeName);
    };
    // 获取当前工程下所有的元数据
    // 获取当前工程下所有的元数据
    /**
     * @param {?} typeName
     * @return {?}
     */
    MetadataService.prototype.GetMetadataListInBo = 
    // 获取当前工程下所有的元数据
    /**
     * @param {?} typeName
     * @return {?}
     */
    function (typeName) {
        return this.gspMetadataService.GetMetadataListInBo(this.currentProjectPath, typeName);
    };
    /**
     * @param {?} typeCode
     * @param {?} postfix
     * @param {?} filter
     * @return {?}
     */
    MetadataService.prototype.GetMetadataByFilter = /**
     * @param {?} typeCode
     * @param {?} postfix
     * @param {?} filter
     * @return {?}
     */
    function (typeCode, postfix, filter) {
        return this.gspMetadataService.GetMetadataByFilter(this.currentProjectPath, typeCode, filter, postfix);
    };
    // 获取元数据工程信息（获取本工程依赖的元数据包基本信息）
    // 获取元数据工程信息（获取本工程依赖的元数据包基本信息）
    /**
     * @return {?}
     */
    MetadataService.prototype.GetMetadataProjectInfo = 
    // 获取元数据工程信息（获取本工程依赖的元数据包基本信息）
    /**
     * @return {?}
     */
    function () {
        return this.gspMetadataService.GetProjectInfo(this.currentProjectPath);
    };
    // 获取工程路径
    // 获取工程路径
    /**
     * @param {?} metadataPath
     * @return {?}
     */
    MetadataService.prototype.GetMetadataProjPath = 
    // 获取工程路径
    /**
     * @param {?} metadataPath
     * @return {?}
     */
    function (metadataPath) {
        return this.gspMetadataService.GetProjectInfo(metadataPath);
    };
    // tslint:disable-next-line:max-line-length
    // tslint:disable-next-line:max-line-length
    /**
     * @param {?} scope
     * @param {?} codelanguage
     * @param {?} isMavenEnable
     * @param {?=} metadataType
     * @param {?=} searchText
     * @param {?=} page
     * @param {?=} selectedPackageSource
     * @return {?}
     */
    MetadataService.prototype.GetMetadataInfos = 
    // tslint:disable-next-line:max-line-length
    /**
     * @param {?} scope
     * @param {?} codelanguage
     * @param {?} isMavenEnable
     * @param {?=} metadataType
     * @param {?=} searchText
     * @param {?=} page
     * @param {?=} selectedPackageSource
     * @return {?}
     */
    function (scope, codelanguage, isMavenEnable, metadataType, searchText, page, selectedPackageSource) {
        if (metadataType === void 0) { metadataType = null; }
        if (searchText === void 0) { searchText = null; }
        if (page === void 0) { page = null; }
        if (selectedPackageSource === void 0) { selectedPackageSource = null; }
        if (codelanguage === 'dotnet' || isMavenEnable === false) {
            return this.gspMetadataService.GetMetadataInfosByScope(scope, metadataType, searchText, page, selectedPackageSource, this.currentProjectPath);
        }
        else {
            return this.gspMetadataService.GetMetadataInfosByScope4Java(scope, metadataType, searchText, page, selectedPackageSource, this.currentProjectPath);
        }
    };
    // 获取元数据包中所包含的元数据信息
    // 获取元数据包中所包含的元数据信息
    /**
     * @param {?} packPath
     * @param {?} packageName
     * @return {?}
     */
    MetadataService.prototype.GetMetadataPackageInfo = 
    // 获取元数据包中所包含的元数据信息
    /**
     * @param {?} packPath
     * @param {?} packageName
     * @return {?}
     */
    function (packPath, packageName) {
        /** @type {?} */
        var fullName = '';
        if (packageName.endsWith('mdpkg')) {
            fullName = packageName;
        }
        else {
            fullName = packageName + '.mdpkg';
        }
        return this.gspMetadataService.GetMetadataPackageInfo(packPath, fullName);
    };
    /**
     * @param {?} scope
     * @param {?} metadataIndexDto
     * @param {?} codelanguage
     * @param {?} isMavenEnable
     * @return {?}
     */
    MetadataService.prototype.pickMetadataIndex = /**
     * @param {?} scope
     * @param {?} metadataIndexDto
     * @param {?} codelanguage
     * @param {?} isMavenEnable
     * @return {?}
     */
    function (scope, metadataIndexDto, codelanguage, isMavenEnable) {
        // if (codelanguage === 'dotnet' || isMavenEnable === false) {
        //   scope = scope === 3 ? 2 : scope;
        //   return this.gspMetadataService.pickMetadataIndex(scope, this.currentProjectPath, metadataIndexDto);
        // } else {
        return this.gspMetadataService.pickMetadataIndex(scope, this.currentProjectPath, metadataIndexDto);
        // }
    };
    // Load元数据
    // Load元数据
    /**
     * @param {?} metadata
     * @return {?}
     */
    MetadataService.prototype.LoadTheMetadata = 
    // Load元数据
    /**
     * @param {?} metadata
     * @return {?}
     */
    function (metadata) {
        if (metadata) {
            /** @type {?} */
            var fileName = metadata.fileName;
            /** @type {?} */
            var relativePath = metadata.relativePath;
            return this.gspMetadataService.LoadMetadataInBo(fileName, relativePath, this.currentProjectPath);
        }
        return new Observable$1();
    };
    // Load元数据
    // Load元数据
    /**
     * @return {?}
     */
    MetadataService.prototype.LoadMetadata = 
    // Load元数据
    /**
     * @return {?}
     */
    function () {
        if (this.metadatas.length > 0) {
            /** @type {?} */
            var index1 = this.metadatas.length;
            if (index1 > 0) {
                /** @type {?} */
                var index2 = this.metadatas[index1 - 1].treenode.MetadataInfo.length;
                if (index2 > 0) {
                    /** @type {?} */
                    var fileName = this.metadatas[index1 - 1].treenode.MetadataInfo[index2 - 1].FileName;
                    /** @type {?} */
                    var relativePath = this.metadatas[index1 - 1].treenode.MetadataInfo[index2 - 1].RelativePath;
                    return this.gspMetadataService.LoadMetadata(fileName, relativePath);
                }
            }
            else {
                return new Observable$1();
            }
        }
    };
    // // 获取元数据以及包的信息
    // GetTheMetadataFromPackage(metadata): Observable<any> {
    //   if (metadata) {
    //     const packageName = metadata.mdPkgName;
    //     const metadataID = metadata.id;
    //     return this.gspMetadataService.GetMetadataFromPackage(packageName, null, metadataID);
    //   }
    //   return new Observable<any>();
    // }
    // // 获取元数据以及包的信息
    // GetMetadataFromPackage(): Observable<any> {
    //   const index1 = this.metadatas.length;
    //   if (index1 > 0) {
    //     const index2 = this.metadatas[index1 - 1].treenode.MetadataInfo.length;
    //     if (index2 > 0) {
    //       const packageName = this.metadatas[index1 - 1].treenode.PackageName;
    //       const packagePath = this.metadatas[index1 - 1].treenode.Location;
    //       const metadataID = this.metadatas[index1 - 1].treenode.MetadataInfo[index2 - 1].ID;
    //       return this.gspMetadataService.GetMetadataFromPackage(packageName, packagePath, metadataID);
    //     }
    //   } else {
    //     return new Observable<any>();
    //   }
    // }
    // 获取某个路径下的所有元数据包
    // // 获取元数据以及包的信息
    // GetTheMetadataFromPackage(metadata): Observable<any> {
    //   if (metadata) {
    //     const packageName = metadata.mdPkgName;
    //     const metadataID = metadata.id;
    //     return this.gspMetadataService.GetMetadataFromPackage(packageName, null, metadataID);
    //   }
    //   return new Observable<any>();
    // }
    // // 获取元数据以及包的信息
    // GetMetadataFromPackage(): Observable<any> {
    //   const index1 = this.metadatas.length;
    //   if (index1 > 0) {
    //     const index2 = this.metadatas[index1 - 1].treenode.MetadataInfo.length;
    //     if (index2 > 0) {
    //       const packageName = this.metadatas[index1 - 1].treenode.PackageName;
    //       const packagePath = this.metadatas[index1 - 1].treenode.Location;
    //       const metadataID = this.metadatas[index1 - 1].treenode.MetadataInfo[index2 - 1].ID;
    //       return this.gspMetadataService.GetMetadataFromPackage(packageName, packagePath, metadataID);
    //     }
    //   } else {
    //     return new Observable<any>();
    //   }
    // }
    // 获取某个路径下的所有元数据包
    /**
     * @return {?}
     */
    MetadataService.prototype.GetMetadataPackages = 
    // // 获取元数据以及包的信息
    // GetTheMetadataFromPackage(metadata): Observable<any> {
    //   if (metadata) {
    //     const packageName = metadata.mdPkgName;
    //     const metadataID = metadata.id;
    //     return this.gspMetadataService.GetMetadataFromPackage(packageName, null, metadataID);
    //   }
    //   return new Observable<any>();
    // }
    // // 获取元数据以及包的信息
    // GetMetadataFromPackage(): Observable<any> {
    //   const index1 = this.metadatas.length;
    //   if (index1 > 0) {
    //     const index2 = this.metadatas[index1 - 1].treenode.MetadataInfo.length;
    //     if (index2 > 0) {
    //       const packageName = this.metadatas[index1 - 1].treenode.PackageName;
    //       const packagePath = this.metadatas[index1 - 1].treenode.Location;
    //       const metadataID = this.metadatas[index1 - 1].treenode.MetadataInfo[index2 - 1].ID;
    //       return this.gspMetadataService.GetMetadataFromPackage(packageName, packagePath, metadataID);
    //     }
    //   } else {
    //     return new Observable<any>();
    //   }
    // }
    // 获取某个路径下的所有元数据包
    /**
     * @return {?}
     */
    function () {
        return this.gspMetadataService.GetMetadataPackages('packages');
    };
    // 更新工程引用
    // 更新工程引用
    /**
     * @param {?} packageHeader
     * @return {?}
     */
    MetadataService.prototype.UpdateMetadataProject = 
    // 更新工程引用
    /**
     * @param {?} packageHeader
     * @return {?}
     */
    function (packageHeader) {
        // this.gspMetadataService.UpdateMetadataProject(this.currentProjectPath, packageHeader).then(response => {
        //   if (response.status === 200) {
        //     this.treeData = [];
        //     this.Refresh();
        //   }
        // });
        return this.gspMetadataService.UpdateMetadataProject(this.currentProjectPath, packageHeader);
    };
    /**
     * @return {?}
     */
    MetadataService.prototype.getCodeLanguage = /**
     * @return {?}
     */
    function () {
        return this.gspMetadataService.getCodeLanguage(this.currentProjectPath);
    };
    /**
     * @return {?}
     */
    MetadataService.prototype.getProcessMode = /**
     * @return {?}
     */
    function () {
        return this.gspMetadataService.getProcessMode(this.currentProjectPath);
    };
    /**
     * @return {?}
     */
    MetadataService.prototype.Refresh = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.treeData = this.getMetadataClassifier('');
        this.treeData[1].items = [];
        /** @type {?} */
        var packages = [];
        this.GetMetadataProjPath(this.currentProjectPath).subscribe((/**
         * @param {?} proj
         * @return {?}
         */
        function (proj) {
            _this.currentProjectPath = proj.projectPath;
            _this.GetMetadataProjectInfo().subscribe((/**
             * @param {?} meta
             * @return {?}
             */
            function (meta) {
                packages = meta.nugetPackageRefs;
                packages.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    /** @type {?} */
                    var child = {
                        id: item.name,
                        text: item.name,
                        items: []
                    };
                    if (_this.treeData) {
                        _this.treeData[1].items.push(child);
                    }
                    _this.gspMetadataService.GetDevRootPath().subscribe((/**
                     * @param {?} root
                     * @return {?}
                     */
                    function (root) {
                        /** @type {?} */
                        var rootPath = root + _this.packRefPath;
                        _this.gspMetadataService.GetMetadataPackagesUnderNugetDir(rootPath + '/' + item.name).subscribe((/**
                         * @param {?} pkgs
                         * @return {?}
                         */
                        function (pkgs) {
                            if (pkgs) {
                                /** @type {?} */
                                var mdpkg_1;
                                pkgs.forEach((/**
                                 * @param {?} pkg
                                 * @return {?}
                                 */
                                function (pkg) {
                                    mdpkg_1 = {
                                        id: pkg.header.name,
                                        text: pkg.header.name,
                                        desc: pkg.header.location
                                    };
                                    child.items.push(mdpkg_1);
                                }));
                                // this.files[1].items[index].push(mdpkg);
                            }
                        }));
                    }));
                }));
            }));
        }));
    };
    /**
     * @return {?}
     */
    MetadataService.prototype.getBoPath = /**
     * @return {?}
     */
    function () {
        return this.gspMetadataService.getBoPath(this.currentProjectPath);
    };
    MetadataService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MetadataService.ctorParameters = function () { return [
        { type: GSPMetadataService }
    ]; };
    return MetadataService;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/treeclassifier/treeclassifier.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var TreeclassifierComponent = /** @class */ (function () {
    function TreeclassifierComponent(metadataService, gspMetadataService) {
        this.metadataService = metadataService;
        this.gspMetadataService = gspMetadataService;
        this.selectedKeys = ['0'];
        this.classifier = [
            { id: 'project', text: '当前工程' }
        ];
    }
    /**
     * @return {?}
     */
    TreeclassifierComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.metadataService.GetMetadataProjPath(this.metadataPath).subscribe((/**
         * @param {?} proj
         * @return {?}
         */
        function (proj) {
            _this.metadataService.currentProjectPath = proj.projectPath;
            if (_this.onlyDisplayCurrentProject === true) {
                _this.files = _this.classifier;
                _this.metadataService.selectedChange.emit(_this.files[0]); // 默认本工程下的
            }
            else {
                // 树节点mock数据
                _this.gspMetadataService.getMavenSetting().subscribe((/**
                 * @param {?} maven
                 * @return {?}
                 */
                function (maven) {
                    if (maven === true) {
                        _this.gspMetadataService.getCodeLanguage(proj.projectPath).subscribe((/**
                         * @param {?} lang
                         * @return {?}
                         */
                        function (lang) {
                            if (lang === 'java') {
                                _this.files = _this.metadataService.getMetadataClassifier('bo');
                                _this.metadataService.selectedChange.emit(_this.files[0]); // 默认本工程下的
                            }
                            else {
                                _this.files = _this.metadataService.getMetadataClassifier('project');
                                _this.metadataService.selectedChange.emit(_this.files[0]); // 默认本工程下的
                            }
                        }));
                    }
                    else {
                        _this.files = _this.metadataService.getMetadataClassifier('project');
                        _this.metadataService.selectedChange.emit(_this.files[0]); // 默认本工程下的
                    }
                }));
            }
        }));
    };
    // 树节点点击后事件
    // 树节点点击后事件
    /**
     * @param {?} event
     * @return {?}
     */
    TreeclassifierComponent.prototype.nodeSelect = 
    // 树节点点击后事件
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.metadataService.selectedChange.emit(event.dataItem);
    };
    TreeclassifierComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'app-treeclassifier',
                    template: "<div class=\"templateTree befTree\">\r\n    <div class=\"treeClassifier\">\r\n        <kendo-treeview\r\n            [nodes]=\"files\"\r\n            textField=\"text\"\r\n            kendoTreeViewExpandable\r\n            kendoTreeViewSelectable\r\n            kendoTreeViewHierarchyBinding\r\n            childrenField=\"items\"\r\n            (selectionChange)=\"this.nodeSelect($event)\"\r\n            [(selectedKeys)]=\"this.selectedKeys\"\r\n        >\r\n        </kendo-treeview>\r\n    </div>\r\n</div>",
                    styles: [".treeClassifier .k-treeview{overflow:visible}.templateTree{padding:5px;margin-left:5px}"]
                }] }
    ];
    /** @nocollapse */
    TreeclassifierComponent.ctorParameters = function () { return [
        { type: MetadataService },
        { type: GSPMetadataService }
    ]; };
    TreeclassifierComponent.propDecorators = {
        metadataPath: [{ type: Input }],
        onlyDisplayCurrentProject: [{ type: Input }]
    };
    return TreeclassifierComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/metadatapackageselect/metadatapackageselect.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MetadatapackageselectComponent = /** @class */ (function () {
    function MetadatapackageselectComponent(metadataService, gspMetaService) {
        this.metadataService = metadataService;
        this.gspMetaService = gspMetaService;
        this.closeModalEmitter = new EventEmitter();
    }
    /**
     * @return {?}
     */
    MetadatapackageselectComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    MetadatapackageselectComponent.prototype.showModal = /**
     * @return {?}
     */
    function () {
        // this.isModalShown = true;
    };
    /**
     * @return {?}
     */
    MetadatapackageselectComponent.prototype.hideModal = /**
     * @return {?}
     */
    function () {
        // this.isModalShown = false;
        this.closeModalEmitter.emit();
    };
    /**
     * @return {?}
     */
    MetadatapackageselectComponent.prototype.onHidden = /**
     * @return {?}
     */
    function () {
        // this.isModalShown = false;
        this.closeModalEmitter.emit();
    };
    /**
     * @param {?} classifier
     * @return {?}
     */
    MetadatapackageselectComponent.prototype.onSelectedChange = /**
     * @param {?} classifier
     * @return {?}
     */
    function (classifier) {
        // this.classifier = classifier;
    };
    /**
     * @return {?}
     */
    MetadatapackageselectComponent.prototype.ResolvePackage = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.closeModalEmitter.emit();
        if (this.metadataService.packages.length > 0) {
            /** @type {?} */
            var flag_1 = false;
            /** @type {?} */
            var packageName_1 = this.metadataService.packages[0];
            this.gspMetaService.GetDevRootPath().subscribe((/**
             * @param {?} root
             * @return {?}
             */
            function (root) {
                _this.metadataService.rootPath = root;
                _this.metadataService.GetMetadataPackageInfo(root + _this.metadataService.packRefPath, packageName_1).subscribe((/**
                 * @param {?} packageInfo
                 * @return {?}
                 */
                function (packageInfo) {
                    /** @type {?} */
                    var header = packageInfo.header;
                    // 检测依赖项
                    _this.metadataService.GetMetadataProjectInfo().subscribe((/**
                     * @param {?} projInfo
                     * @return {?}
                     */
                    function (projInfo) {
                        // 已有依赖关系
                        /** @type {?} */
                        var refs = projInfo.metadataPackageRefs;
                        refs.forEach((/**
                         * @param {?} item
                         * @return {?}
                         */
                        function (item) {
                            if (item.name === header.name) {
                                alert('已添加对该元数据包的依赖');
                                flag_1 = true;
                            }
                        }));
                        if (flag_1 === true) {
                            return;
                        }
                        // 更新工程引用
                        _this.metadataService.UpdateMetadataProject(header).subscribe((/**
                         * @param {?} data
                         * @return {?}
                         */
                        function (data) {
                            _this.metadataService.treeData = [];
                            _this.metadataService.Refresh();
                        }));
                        // 刷新树节点
                        if (_this.treeclassifierComponent) {
                            _this.treeclassifierComponent.files = _this.metadataService.treeData;
                        }
                    }));
                }));
            }));
        }
    };
    MetadatapackageselectComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'app-metadatapackageselect',
                    template: "<!--\u7B2C\u4E8C\u79CD\u5F39\u51FA\u65B9\u6CD5\u7684\u7EC4\u4EF6-->\r\n<!-- <div *ngIf=\"isModalShown\" [config]=\"{ show: true }\" (onHidden)=\"onHidden()\" bsModal #autoShownModal=\"bs-modal\" class=\"modal fade\" tabindex=\"-1\" role=\"dialog\" aria-hidden=\"true\" style='min-height: 800px; min-width: 900px; max-width: none'>\r\n    <div class=\"modal-dialog\" style='max-width:none; width: 615px;margin-top:0'>\r\n        <div class=\"modal-content\"> -->\r\n            <!-- <div class=\"modal-header\">\r\n                <h4 class=\"modal-title pull-left\">\u9009\u62E9\u5143\u6570\u636E\u5305</h4>\r\n                <button type=\"button\" class=\"close pull-right\" aria-label=\"Close\" (click)=\"hideModal()\">\r\n          <span aria-hidden=\"true\">&times;</span>\r\n        </button>\r\n            </div> -->\r\n            <div class=\"modal-body\">\r\n                <div class=\"templateBlock\">\r\n                    <app-metadatapackagelist></app-metadatapackagelist>\r\n                </div>\r\n            </div>\r\n            <div class=\"modal-footer\">\r\n                <button type=\"button\" class=\"btn btn-default btn-primary\" (click)=\"ResolvePackage()\">\u786E\u5B9A</button>\r\n                <button type=\"button\" class=\"btn btn-default\" (click)=\"hideModal()\">\u53D6\u6D88</button>\r\n            </div>\r\n        <!-- </div>\r\n    </div>\r\n</div> -->",
                    styles: [".templateBlock{display:flex}.infoBlock{width:100%;min-width:400px}.infoBlock label{width:50px}"]
                }] }
    ];
    /** @nocollapse */
    MetadatapackageselectComponent.ctorParameters = function () { return [
        { type: MetadataService },
        { type: GSPMetadataService }
    ]; };
    MetadatapackageselectComponent.propDecorators = {
        closeModalEmitter: [{ type: Output }],
        treeclassifierComponent: [{ type: ViewChild, args: [TreeclassifierComponent,] }]
    };
    return MetadatapackageselectComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata4refdto.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var Metadata4RefDto = /** @class */ (function () {
    function Metadata4RefDto() {
    }
    return Metadata4RefDto;
}());
var MetadataPackageHeader = /** @class */ (function () {
    function MetadataPackageHeader() {
    }
    return MetadataPackageHeader;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/selectmetadate/selectmetadata.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SelectMetadataComponent = /** @class */ (function () {
    function SelectMetadataComponent(metadataService, loading, gspMetaService) {
        var _this = this;
        this.metadataService = metadataService;
        this.loading = loading;
        this.gspMetaService = gspMetaService;
        this.selectedMetadata = new EventEmitter();
        this.closeModalEmitter = new EventEmitter();
        this.metadataService.selectedChange.subscribe((/**
         * @param {?} leftItem
         * @return {?}
         */
        function (leftItem) {
            _this.classifier = leftItem;
        }));
        this.isMavenEnable = false;
        this.gspMetaService.getMavenSetting().subscribe((/**
         * @param {?} flag
         * @return {?}
         */
        function (flag) {
            // if (flag) {
            _this.isMavenEnable = flag;
            // }
        }));
    }
    /**
     * @param {?} metadata
     * @return {?}
     */
    SelectMetadataComponent.prototype.onSelectedMetadataChanged = /**
     * @param {?} metadata
     * @return {?}
     */
    function (metadata) {
        this.metadata = metadata;
    };
    /**
     * @param {?} code
     * @return {?}
     */
    SelectMetadataComponent.prototype.setCodeLanguage = /**
     * @param {?} code
     * @return {?}
     */
    function (code) {
        this.codelanguage = code;
    };
    /**
     * @return {?}
     */
    SelectMetadataComponent.prototype.LoadMetadata = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.metadata && this.classifier) {
            // 选中的是当前工程下的元数据，调用Load
            if (this.classifier.id === 'project' || this.classifier.id === 'bo') {
                if (this.loading) {
                    this.loading.show({ message: '正在添加引用，请稍候...' });
                }
                this.metadataService.LoadTheMetadata(this.metadata).subscribe((/**
                 * @param {?} templates
                 * @return {?}
                 */
                function (templates) {
                    if (_this.loading) {
                        _this.loading.close();
                    }
                    /** @type {?} */
                    var metadata4Ref = new Metadata4RefDto();
                    metadata4Ref.metadata = templates;
                    metadata4Ref.metadata.processMode = _this.metadata.processMode === 'interpretation' ? 'interpretation' : 'generation';
                    metadata4Ref.packageHeader = new MetadataPackageHeader();
                    _this.selectedMetadata.emit(metadata4Ref);
                    _this.hideModal();
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    _this.errorMsg(error);
                    if (_this.loading) {
                        _this.loading.close();
                    }
                }), (/**
                 * @return {?}
                 */
                function () {
                    if (_this.loading) {
                        _this.loading.close();
                    }
                }));
            }
            else if (this.classifier.id === 'local') {
                if (!this.metadata.mdPkgName) {
                    /** @type {?} */
                    var msg = '未找到元数据包信息。';
                    gsp.ide.messager.error(msg);
                    return;
                }
                if (this.loading) {
                    this.loading.show({ message: '正在添加引用，请稍候...' });
                }
                this.metadataService.pickMetadataIndex(1, this.metadata, this.codelanguage, this.isMavenEnable).subscribe((/**
                 * @param {?} templates
                 * @return {?}
                 */
                function (templates) {
                    if (_this.loading) {
                        _this.loading.close();
                    }
                    templates.metadata.processMode = _this.metadata.processMode === 'interpretation' ? 'interpretation' : 'generation';
                    _this.selectedMetadata.emit(templates);
                    _this.hideModal();
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    _this.errorMsg(error);
                    if (_this.loading) {
                        _this.loading.close();
                    }
                }), (/**
                 * @return {?}
                 */
                function () {
                    if (_this.loading) {
                        _this.loading.close();
                    }
                }));
            }
            else if (this.classifier.id === 'remote') {
                if (!this.metadata.mdPkgName) {
                    /** @type {?} */
                    var msg = '未找到元数据包信息。';
                    gsp.ide.messager.error(msg);
                    return;
                }
                if (!this.metadata.repoPkg || this.metadata.repoPkg.length < 1) {
                    /** @type {?} */
                    var msg = '未找到仓库信息。';
                    gsp.ide.messager.error(msg);
                    return;
                }
                if (this.loading) {
                    this.loading.show({ message: '正在添加引用，请稍候...' });
                }
                this.metadataService.pickMetadataIndex(2, this.metadata, this.codelanguage, this.isMavenEnable).subscribe((/**
                 * @param {?} templates
                 * @return {?}
                 */
                function (templates) {
                    if (_this.loading) {
                        _this.loading.close();
                    }
                    templates.metadata.processMode = _this.metadata.processMode === 'interpretation' ? 'interpretation' : 'generation';
                    _this.selectedMetadata.emit(templates);
                    _this.hideModal();
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    _this.errorMsg(error);
                    if (_this.loading) {
                        _this.loading.close();
                    }
                }), (/**
                 * @return {?}
                 */
                function () {
                    if (_this.loading) {
                        _this.loading.close();
                    }
                }));
            }
            else if (this.classifier.id === 'ref') {
                if (!this.metadata.mdPkgName) {
                    /** @type {?} */
                    var msg = '未找到元数据包信息。';
                    gsp.ide.messager.error(msg);
                    return;
                }
                if (!this.metadata.repoPkg || this.metadata.repoPkg.length < 1) {
                    /** @type {?} */
                    var msg = '未找到仓库信息。';
                    gsp.ide.messager.error(msg);
                    return;
                }
                if (this.loading) {
                    this.loading.show({ message: '请稍候...' });
                }
                this.metadataService.pickMetadataIndex(3, this.metadata, this.codelanguage, this.isMavenEnable).subscribe((/**
                 * @param {?} templates
                 * @return {?}
                 */
                function (templates) {
                    if (_this.loading) {
                        _this.loading.close();
                    }
                    templates.metadata.processMode = _this.metadata.processMode === 'interpretation' ? 'interpretation' : 'generation';
                    _this.selectedMetadata.emit(templates);
                    _this.hideModal();
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    _this.errorMsg(error);
                    if (_this.loading) {
                        _this.loading.close();
                    }
                }), (/**
                 * @return {?}
                 */
                function () {
                    if (_this.loading) {
                        _this.loading.close();
                    }
                }));
            }
        }
        else {
            /** @type {?} */
            var msg = '请先选中一个元数据。';
            gsp.ide.messager.warning(msg);
        }
    };
    /**
     * @private
     * @param {?} error
     * @return {?}
     */
    SelectMetadataComponent.prototype.errorMsg = /**
     * @private
     * @param {?} error
     * @return {?}
     */
    function (error) {
        /** @type {?} */
        var msg = error.error.Message !== undefined ? error.error.Message : error.error.error.message;
        if (msg == null) {
            msg = '引用元数据失败。';
        }
        gsp.ide.messager.error(msg);
    };
    // 弹出元数据包选择界面
    // 弹出元数据包选择界面
    /**
     * @return {?}
     */
    SelectMetadataComponent.prototype.PackageSelect = 
    // 弹出元数据包选择界面
    /**
     * @return {?}
     */
    function () {
        this.metadatapackageselectComponent.showModal();
    };
    /**
     * @return {?}
     */
    SelectMetadataComponent.prototype.hideModal = /**
     * @return {?}
     */
    function () {
        this.closeModalEmitter.emit();
    };
    SelectMetadataComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'app-select-metadata',
                    template: "<div class=\"farris-main-area d-flex flex-column\">\r\n  <div class=\"modal-body d-flex flex-fill flex-column\">\r\n    <div class=\"templateBlock flex-column flex-fill d-flex\">\r\n      <kendo-splitter class=\"d-flex \" style=\"flex: 1 1 auto;\" orientation=\"horizontal\">\r\n        <kendo-splitter-pane class=\"d-flex \" [collapsible]=\"true\" size=\"170px\" style=\"height: auto\">\r\n          <app-treeclassifier [metadataPath]=\"metadataPath\" [onlyDisplayCurrentProject]=\"onlyDisplayCurrentProject\"></app-treeclassifier>\r\n        </kendo-splitter-pane>\r\n        <kendo-splitter-pane class=\"d-flex\" style=\"height: auto\">\r\n          <app-metadatalist [typeName]=\"typeName\" [filter]=\"filter\" [isMavenEnable]=\"isMavenEnable\" (technologyStack)=\"setCodeLanguage($event)\"\r\n            (selectedMetadataChanged)=\"this.onSelectedMetadataChanged($event)\"></app-metadatalist>\r\n        </kendo-splitter-pane>\r\n      </kendo-splitter>\r\n    </div>\r\n  </div>\r\n  <div class=\"modal-footer\">\r\n    <button type=\"button\" click.once class=\"btn btn-outline-primary btn-sm px-3 mx-2\"\r\n      (click.once)=\"LoadMetadata()\">\u786E\u5B9A</button>\r\n    <button type=\"button\" class=\"btn btn-outline-secondary btn-sm px-3 mx-2\" (click)=\"hideModal()\">\u53D6\u6D88</button>\r\n  </div>\r\n</div>\r\n",
                    styles: [".templateBlock{display:flex}.infoBlock{width:100%;min-width:400px}.infoBlock label{width:50px}"]
                }] }
    ];
    /** @nocollapse */
    SelectMetadataComponent.ctorParameters = function () { return [
        { type: MetadataService },
        { type: LoadingService },
        { type: GSPMetadataService }
    ]; };
    SelectMetadataComponent.propDecorators = {
        metadatapackageselectComponent: [{ type: ViewChild, args: [MetadatapackageselectComponent,] }],
        metadataPath: [{ type: Input }],
        typeName: [{ type: Input }],
        filter: [{ type: Input }],
        onlyDisplayCurrentProject: [{ type: Input }],
        closeModalEmitter: [{ type: Output }],
        selectedMetadata: [{ type: Output }]
    };
    return SelectMetadataComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/metadatalist/metadatalist.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var TemplatemetadataComponent = /** @class */ (function () {
    function TemplatemetadataComponent(metadataService, gspMetaService) {
        var _this = this;
        this.metadataService = metadataService;
        this.gspMetaService = gspMetaService;
        this.subscriptions = [];
        this.dotnet = 'dotnet';
        this.java = 'java';
        this.technologyStack = new EventEmitter();
        this.loading = false;
        this.selectedMetadataChanged = new EventEmitter();
        this.allSource = { text: '全选', value: 'All' };
        this.selectedSource = this.allSource;
        this.sourceList = [this.allSource];
        this.popupSettings = { width: 'auto' };
        this.allowCustom = true;
        this.allType = { text: '全选', value: 'All' };
        this.selectedType = this.allType;
        this.typeList = [
            this.allType
        ];
        this.postfixSelected = '';
        this.initPageIndex = 0;
        this.initPageSize = 10;
        this.page = {
            pageIndex: this.initPageIndex,
            pageSize: this.initPageSize
        };
        this.typeListReadonly = false;
        this.hasNext = true;
        this.searchText = '';
        this.searchTextForDup = '';
        this.searchTextSub = new Subject();
        this.postfix = '';
        this.postfixFlag = true;
        this.scope = 0;
        this.conflictAvoidFlag = false;
        this.reload = true;
        this.mySelection = [];
        // tslint:disable-next-line:member-ordering
        this.state = {
            skip: 0,
            take: 9999,
            // Initial filter descriptor
            filter: {
                logic: 'and',
                filters: []
            }
        };
        // tslint:disable-next-line:member-ordering
        this.selectableSetting = { mode: 'single', checkboxOnly: false };
        this.rowSelected = (/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return _this.mySelection.indexOf(e.index) >= 0; });
        this.classifier = { id: '' };
        this.searchTextSub.pipe(debounceTime(2048), distinctUntilChanged()).subscribe((/**
         * @param {?} text
         * @return {?}
         */
        function (text) {
            _this.searchText = text.trim();
            // 搜索
            console.log('search for ' + _this.searchText);
            _this.getData();
        }));
        // this.unsubscribeAll();
        /** @type {?} */
        var subscription = this.metadataService.selectedChange.subscribe((/**
         * @param {?} classifier
         * @return {?}
         */
        function (classifier) {
            if (_this.classifier && classifier && _this.classifier.id === classifier.id) {
                return;
            }
            // 元数据类型
            if (_this.postfixFlag) {
                _this.gspMetaService.GetMetadataTypeList().subscribe((/**
                 * @param {?} typeList
                 * @return {?}
                 */
                function (typeList) {
                    /** @type {?} */
                    var count = 0;
                    typeList.forEach((/**
                     * @param {?} type
                     * @return {?}
                     */
                    function (type) {
                        _this.postfixFlag = false;
                        /** @type {?} */
                        var typeNameArray = _this.typeName.split(',');
                        typeNameArray.forEach((/**
                         * @param {?} typeName
                         * @return {?}
                         */
                        function (typeName) {
                            if (typeName.toLowerCase() === type.typeCode.toLowerCase()) {
                                _this.postfix = _this.postfix + type.postfix + ',';
                                _this.typeList.push({ text: type.typeName, value: type.postfix });
                                count++;
                            }
                        }));
                    }));
                    if (count === 1) {
                        _this.typeList.shift();
                        _this.selectedType = _this.typeList[0];
                        _this.typeListReadonly = true;
                    }
                    _this.postfixSelected = _this.postfix;
                    _this.scopeChanged(classifier);
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    /** @type {?} */
                    var msg = error.error.Message !== undefined ? error.error.Message : error.error.error.message;
                    gsp.ide.messager.error('Fetch Metadata Type error:' + msg);
                }));
            }
            else {
                _this.scopeChanged(classifier);
            }
            // 技术栈赋值
            _this.metadataService.getCodeLanguage().subscribe((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item) {
                    // 如果目前工程非单栈，则获取环境默认技术栈
                    if (item.indexOf(';') > -1) {
                        _this.gspMetaService.getSettingInfo().subscribe((/**
                         * @param {?} setting
                         * @return {?}
                         */
                        function (setting) {
                            if (setting && setting.technologyStack) {
                                _this.codelanguage = setting.technologyStack;
                                _this.technologyStack.emit(_this.codelanguage);
                                if (classifier.id === 'ref' || classifier.id === 'remote') {
                                    if (_this.codelanguage === 'dotnet' || _this.isMavenEnable === false) {
                                        _this.initNugetSourceList();
                                    }
                                    else {
                                        _this.initMavenSourceList();
                                    }
                                }
                            }
                            else if (_this.isMavenEnable === true) {
                                // 环境默认配置为空，则按照java技术栈处理
                                _this.codelanguage = 'java';
                                _this.technologyStack.emit(_this.codelanguage);
                                if (classifier.id === 'ref' || classifier.id === 'remote') {
                                    _this.initMavenSourceList();
                                }
                            }
                            else {
                                _this.codelanguage = 'dotnet';
                                _this.technologyStack.emit(_this.codelanguage);
                                if (classifier.id === 'ref' || classifier.id === 'remote') {
                                    _this.initNugetSourceList();
                                }
                            }
                        }));
                    }
                    else {
                        _this.codelanguage = item;
                        _this.technologyStack.emit(_this.codelanguage);
                        if (classifier.id === 'ref' || classifier.id === 'remote') {
                            if (_this.codelanguage === 'dotnet' || _this.isMavenEnable === false) {
                                _this.initNugetSourceList();
                            }
                            else {
                                _this.initMavenSourceList();
                            }
                        }
                    }
                }
                else {
                    // 工程的技术栈为空，则也是从环境默认技术栈读取
                    _this.gspMetaService.getSettingInfo().subscribe((/**
                     * @param {?} setting
                     * @return {?}
                     */
                    function (setting) {
                        if (setting && setting.technologyStack) {
                            _this.codelanguage = setting.technologyStack;
                            _this.technologyStack.emit(_this.codelanguage);
                            if (classifier.id === 'ref' || classifier.id === 'remote') {
                                if (_this.codelanguage === 'dotnet' || _this.isMavenEnable === false) {
                                    _this.initNugetSourceList();
                                }
                                else {
                                    _this.initMavenSourceList();
                                }
                            }
                        }
                        else if (_this.isMavenEnable === true) {
                            // 环境默认配置为空，则按照java技术栈处理
                            _this.codelanguage = 'java';
                            _this.technologyStack.emit(_this.codelanguage);
                            if (classifier.id === 'ref' || classifier.id === 'remote') {
                                _this.initMavenSourceList();
                            }
                        }
                        else {
                            _this.codelanguage = 'dotnet';
                            _this.technologyStack.emit(_this.codelanguage);
                            if (classifier.id === 'ref' || classifier.id === 'remote') {
                                _this.initNugetSourceList();
                            }
                        }
                    }));
                }
            }));
            // 处理模式
            _this.metadataService.getProcessMode().subscribe((/**
             * @param {?} processMode
             * @return {?}
             */
            function (processMode) {
                _this.processMode = processMode;
            }));
        }));
        this.subscriptions.push(subscription);
    }
    /**
     * @return {?}
     */
    TemplatemetadataComponent.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.data = [];
        this.selectionChanged(null, []);
        this.page = {
            pageIndex: this.initPageIndex,
            pageSize: this.initPageSize
        };
    };
    /**
     * @param {?} state
     * @return {?}
     */
    TemplatemetadataComponent.prototype.dataStateChange = /**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this.state = state;
    };
    /**
     * @return {?}
     */
    TemplatemetadataComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        if (this.subscriptions && this.subscriptions.length > 0) {
            try {
                for (var _b = __values(this.subscriptions), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var item = _c.value;
                    item.unsubscribe();
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
    };
    /**
     * @return {?}
     */
    TemplatemetadataComponent.prototype.initNugetSourceList = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.sourceList.length === 1) {
            this.gspMetaService.getNugetSourceList().subscribe((/**
             * @param {?} sources
             * @return {?}
             */
            function (sources) {
                sources.forEach((/**
                 * @param {?} source
                 * @return {?}
                 */
                function (source) {
                    _this.sourceList.push({ text: source.name, value: source.name });
                }));
            }));
        }
    };
    /**
     * @return {?}
     */
    TemplatemetadataComponent.prototype.initMavenSourceList = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.sourceList.length === 1) {
            this.gspMetaService.getMavenSourceList().subscribe((/**
             * @param {?} sources
             * @return {?}
             */
            function (sources) {
                if (sources.repoList) {
                    sources.repoList.forEach((/**
                     * @param {?} source
                     * @return {?}
                     */
                    function (source) {
                        _this.sourceList.push({ text: source.id, value: source.id });
                    }));
                }
            }));
        }
    };
    /**
     * @return {?}
     */
    TemplatemetadataComponent.prototype.onSearchTextChanged = /**
     * @return {?}
     */
    function () {
        if (!this.searchText) {
            this.searchText = '';
        }
        /** @type {?} */
        var searchTextTrim = this.searchText.trim();
        if (this.searchTextForDup === searchTextTrim) {
            return;
        }
        this.searchTextForDup = searchTextTrim;
        console.log(this.searchText);
        this.searchTextSub.next(searchTextTrim);
    };
    /**
     * @private
     * @return {?}
     */
    TemplatemetadataComponent.prototype.GetMetadataList = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var metadatadto = [];
        this.loading = true;
        this.metadataService.GetMetadataList(this.postfixSelected).subscribe((/**
         * @param {?} templates
         * @return {?}
         */
        function (templates) {
            _this.loading = false;
            if (!templates) {
                _this.gridData = _this.data;
                gsp.ide.messager.error('error to get response');
                return;
            }
            // tslint:disable-next-line: prefer-for-of
            for (var index = 0; index < templates.length; index++) {
                /** @type {?} */
                var template = templates[index];
                if (_this.searchText) {
                    // tslint:disable-next-line: max-line-length
                    if (template.code.indexOf(_this.searchText) > -1 || template.name.indexOf(_this.searchText) > -1) {
                        metadatadto.push(template);
                    }
                }
                else {
                    metadatadto.push(template);
                }
            }
            if (metadatadto.length > 0) {
                _this.onSelectionChangedInternal(metadatadto[0], [0]);
            }
            _this.hasNext = false;
            _this.data = metadatadto;
            _this.gridData = _this.data;
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.data = metadatadto;
            _this.gridData = _this.data;
            _this.loading = false;
            /** @type {?} */
            var msg = error.error.Message !== undefined ? error.error.Message : error.error.error.message;
            gsp.ide.messager.error('Get Metadatas Error:' + msg);
        }));
    };
    /**
     * @private
     * @return {?}
     */
    TemplatemetadataComponent.prototype.GetMetadataListInBo = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var metadatadto = [];
        this.loading = true;
        this.metadataService.GetMetadataListInBo(this.postfixSelected).subscribe((/**
         * @param {?} templates
         * @return {?}
         */
        function (templates) {
            _this.loading = false;
            if (!templates) {
                _this.gridData = _this.data;
                gsp.ide.messager.error('error to get response');
                return;
            }
            // tslint:disable-next-line: prefer-for-of
            for (var index = 0; index < templates.length; index++) {
                /** @type {?} */
                var template = templates[index];
                if (_this.searchText) {
                    // tslint:disable-next-line: max-line-length
                    if (template.projectName.toLowerCase().indexOf(_this.searchText.toLowerCase()) > -1
                        || template.code.toLowerCase().indexOf(_this.searchText.toLowerCase()) > -1
                        || template.name.toLowerCase().indexOf(_this.searchText.toLowerCase()) > -1) {
                        metadatadto.push(template);
                    }
                }
                else {
                    metadatadto.push(template);
                }
            }
            if (metadatadto.length > 0) {
                _this.onSelectionChangedInternal(metadatadto[0], [0]);
            }
            _this.hasNext = false;
            _this.data = metadatadto;
            _this.gridData = _this.data;
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.data = metadatadto;
            _this.gridData = _this.data;
            _this.loading = false;
            /** @type {?} */
            var msg = error.error.Message !== undefined ? error.error.Message : error.error.error.message;
            gsp.ide.messager.error('Get Metadatas Error:' + msg);
        }));
    };
    /**
     * @private
     * @param {?=} reloadT
     * @return {?}
     */
    TemplatemetadataComponent.prototype.GetLocalMetadataList = /**
     * @private
     * @param {?=} reloadT
     * @return {?}
     */
    function (reloadT) {
        var _this = this;
        if (reloadT === void 0) { reloadT = true; }
        /** @type {?} */
        var metadatadto = [];
        this.loading = true;
        // this.gspMetaService.GetDevRootPath().subscribe(root => {
        // this.metadataService.rootPath = root;
        // tslint:disable-next-line:max-line-length
        this.metadataService.GetMetadataInfos(this.scope, this.codelanguage, this.isMavenEnable, this.postfixSelected, this.searchText, this.page).subscribe((/**
         * @param {?} templates
         * @return {?}
         */
        function (templates) {
            _this.loading = false;
            if (!templates) {
                gsp.ide.messager.error('error to get metadata infos');
                return;
            }
            // 页码处理
            if (templates.page) {
                if (templates.page.pageSize < _this.initPageSize) {
                    _this.hasNext = false;
                }
                else {
                    _this.hasNext = true;
                }
                _this.page.pageSize = templates.page.pageSize;
                _this.page.pageIndex = templates.page.pageIndex;
            }
            else {
                _this.hasNext = false;
            }
            /** @type {?} */
            var metadataList = templates.metadataIndexItems;
            if (!metadataList || metadataList.length < 1) {
                _this.mySelection = [];
                _this.gridData = _this.data;
                return;
            }
            metadataList.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                if (element.repoPkg && element.repoPkg.length > 0) {
                    element.nugetPkgName = element.repoPkg[0].name;
                    element.nugetPkgVersion = element.repoPkg[0].version;
                }
                metadatadto.push(element);
            }));
            if (reloadT) {
                _this.onSelectionChangedInternal(metadatadto[0], [0]);
                _this.data = metadatadto;
            }
            else {
                metadatadto.forEach((/**
                 * @param {?} dto
                 * @return {?}
                 */
                function (dto) {
                    _this.data.push(dto);
                }));
            }
            _this.gridData = _this.data;
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.data = metadatadto;
            _this.gridData = _this.data;
            _this.loading = false;
            /** @type {?} */
            var msg = error.error.Message !== undefined ? error.error.Message : error.error.error.message;
            gsp.ide.messager.error('Metadata req failed: ' + msg);
        }));
    };
    /**
     * @param {?} dataItem
     * @param {?} index
     * @return {?}
     */
    TemplatemetadataComponent.prototype.showOnlyBeveragesDetails = /**
     * @param {?} dataItem
     * @param {?} index
     * @return {?}
     */
    function (dataItem, index) {
        return dataItem.repoPkg && dataItem.repoPkg.length > 1;
    };
    /**
     * @param {?} reload
     * @return {?}
     */
    TemplatemetadataComponent.prototype.getRemoteMetadataList = /**
     * @param {?} reload
     * @return {?}
     */
    function (reload) {
        var _this = this;
        /** @type {?} */
        var metadatadto = [];
        this.loading = true;
        // tslint:disable-next-line:max-line-length
        this.metadataService.GetMetadataInfos(this.scope, this.codelanguage, this.isMavenEnable, this.postfixSelected, this.searchText, this.page, this.selectedSource.value).subscribe((/**
         * @param {?} templates
         * @return {?}
         */
        function (templates) {
            _this.loading = false;
            if (!templates) {
                gsp.ide.messager.error('error to get metadata infos');
                return;
            }
            // 页码处理
            if (templates.page) {
                if (templates.page.pageSize < _this.initPageSize) {
                    _this.hasNext = false;
                }
                else {
                    _this.hasNext = true;
                }
                _this.page.pageSize = templates.page.pageSize;
                _this.page.pageIndex = templates.page.pageIndex;
            }
            else {
                _this.hasNext = false;
            }
            /** @type {?} */
            var metadataList = templates.metadataIndexItems;
            if (!metadataList || metadataList.length < 1) {
                _this.mySelection = [];
                _this.gridData = _this.data;
                return;
            }
            metadataList.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                if (!element.repoPkg || element.repoPkg.length < 1) {
                    return;
                }
                if (_this.codelanguage === 'dotnet' || _this.isMavenEnable === false) {
                    element.repoPkg.sort(_this.compareNuget('version'));
                    element.nugetPkgName = element.repoPkg[0].name;
                    element.nugetPkgVersion = element.repoPkg[0].version;
                }
                else {
                    element.repoPkg.sort(_this.compare('version'));
                    element.groupId = element.repoPkg[0].groupId;
                    element.artifactId = element.repoPkg[0].artifactId;
                    element.mavenPkgVersion = element.repoPkg[0].version;
                }
                metadatadto.push(element);
            }));
            if (!metadatadto || metadatadto.length < 1) {
                _this.mySelection = [];
                _this.gridData = _this.data;
                return;
            }
            if (reload) {
                _this.onSelectionChangedInternal(metadatadto[0], [0]);
                _this.data = metadatadto;
            }
            else {
                metadatadto.forEach((/**
                 * @param {?} dto
                 * @return {?}
                 */
                function (dto) {
                    _this.data.push(dto);
                }));
            }
            _this.gridData = _this.data;
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.data = metadatadto;
            _this.gridData = _this.data;
            _this.loading = false;
            /** @type {?} */
            var msg = error.error.Message !== undefined ? error.error.Message : error.error.error.message;
            gsp.ide.messager.error('Metadata req failed: ' + msg);
        }));
    };
    /**
     * @param {?} pro
     * @return {?}
     */
    TemplatemetadataComponent.prototype.compare = /**
     * @param {?} pro
     * @return {?}
     */
    function (pro) {
        return (/**
         * @param {?} obj1
         * @param {?} obj2
         * @return {?}
         */
        function (obj1, obj2) {
            /** @type {?} */
            var val1 = obj1[pro];
            /** @type {?} */
            var val2 = obj2[pro];
            /** @type {?} */
            var vala1 = val1.split('.');
            /** @type {?} */
            var vala2 = val2.split('.');
            /** @type {?} */
            var v10 = vala1[0].substr(1);
            /** @type {?} */
            var v20 = vala2[0].substr(1);
            if (Number(v10) < Number(v20)) {
                return 1;
            }
            if (Number(v10) > Number(v20)) {
                return -1;
            }
            if (Number(vala1[1]) < Number(vala2[1])) {
                return 1;
            }
            if (Number(vala1[1]) > Number(vala2[1])) {
                return -1;
            }
            /** @type {?} */
            var v12 = vala1[2].replace('-SNAPSHOT', '');
            /** @type {?} */
            var v22 = vala2[2].replace('-SNAPSHOT', '');
            if (Number(v12) < Number(v22)) {
                return 1;
            }
            if (Number(v12) > Number(v22)) {
                return -1;
            }
            if (vala1[2].indexOf('-SNAPSHOT') > 0) {
                return 1;
            }
            else {
                return -1;
            }
        });
    };
    /**
     * @param {?} pro
     * @return {?}
     */
    TemplatemetadataComponent.prototype.compareNuget = /**
     * @param {?} pro
     * @return {?}
     */
    function (pro) {
        return (/**
         * @param {?} obj1
         * @param {?} obj2
         * @return {?}
         */
        function (obj1, obj2) {
            /** @type {?} */
            var val1 = obj1[pro];
            /** @type {?} */
            var val2 = obj2[pro];
            /** @type {?} */
            var vala1 = val1.split('.');
            /** @type {?} */
            var vala2 = val2.split('.');
            /** @type {?} */
            var v10 = vala1[0];
            /** @type {?} */
            var v20 = vala2[0];
            if (Number(v10) < Number(v20)) {
                return -1;
            }
            if (Number(v10) > Number(v20)) {
                return 1;
            }
            if (Number(vala1[1]) < Number(vala2[1])) {
                return 1;
            }
            if (Number(vala1[1]) > Number(vala2[1])) {
                return -1;
            }
            /** @type {?} */
            var v12 = vala1[2];
            /** @type {?} */
            var v22 = vala2[2];
            if (v12.indexOf('alpha') > 0) {
                if (v12 < v22) {
                    return 1;
                }
                if (v12 > v22) {
                    return -1;
                }
            }
            else {
                if (Number(v12) < Number(v22)) {
                    return 1;
                }
                if (Number(v12) > Number(v22)) {
                    return -1;
                }
            }
        });
    };
    /**
     * @param {?} classifier
     * @return {?}
     */
    TemplatemetadataComponent.prototype.scopeChanged = /**
     * @param {?} classifier
     * @return {?}
     */
    function (classifier) {
        this.classifier = classifier;
        this.getData();
    };
    /**
     * @return {?}
     */
    TemplatemetadataComponent.prototype.getData = /**
     * @return {?}
     */
    function () {
        if (this.classifier) {
            if (this.reload) {
                this.clear();
            }
            /** @type {?} */
            var reloadT = this.reload;
            if (this.classifier.id === 'project') {
                this.scope = 0;
                this.GetMetadataList();
            }
            else if (this.classifier.id === 'bo') {
                this.scope = 0;
                this.GetMetadataListInBo();
            }
            else if (this.classifier.id === 'ref') {
                this.scope = 3;
                this.getRemoteMetadataList(reloadT);
            }
            else if (this.classifier.id === 'remote') {
                this.scope = 2;
                this.getRemoteMetadataList(reloadT);
            }
            else if (this.classifier.id === 'local') {
                this.scope = 1;
                this.GetLocalMetadataList(reloadT);
            }
            // 重置一次请求中的状态
            this.reload = true;
        }
    };
    // 选中元数据后的事件处理
    // 选中元数据后的事件处理
    /**
     * @param {?} event
     * @return {?}
     */
    TemplatemetadataComponent.prototype.onSelectionChange = 
    // 选中元数据后的事件处理
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event && event.selectedRows && event.selectedRows.length > 0) {
            /** @type {?} */
            var dataItem = event.selectedRows[0].dataItem;
            this.onSelectionChangedInternal(dataItem, [event.selectedRows[0].index]);
        }
    };
    /**
     * @private
     * @param {?} dataItem
     * @param {?} index
     * @return {?}
     */
    TemplatemetadataComponent.prototype.onSelectionChangedInternal = /**
     * @private
     * @param {?} dataItem
     * @param {?} index
     * @return {?}
     */
    function (dataItem, index) {
        /** @type {?} */
        var newData = this.makeNewData(dataItem);
        if (this.scope === 2 || this.scope === 3) {
            newData.repoPkg.push(dataItem.repoPkg[0]);
        }
        this.selectionChanged(newData, index);
    };
    /**
     * @private
     * @param {?} dataItem
     * @param {?=} index
     * @return {?}
     */
    TemplatemetadataComponent.prototype.selectionChanged = /**
     * @private
     * @param {?} dataItem
     * @param {?=} index
     * @return {?}
     */
    function (dataItem, index) {
        if (index === void 0) { index = []; }
        this.selectedMetadataChanged.emit(dataItem);
        this.mySelection = index;
    };
    /**
     * @private
     * @param {?} dataItem
     * @return {?}
     */
    TemplatemetadataComponent.prototype.makeNewData = /**
     * @private
     * @param {?} dataItem
     * @return {?}
     */
    function (dataItem) {
        return {
            mdPkgName: dataItem.mdPkgName,
            repoPkg: [],
            id: dataItem.id,
            nameSpace: dataItem.nameSpace,
            code: dataItem.code,
            name: dataItem.name,
            fileName: dataItem.fileName,
            type: dataItem.type,
            bizobjectID: dataItem.bizobjectID,
            language: dataItem.language,
            isTranslating: dataItem.isTranslating,
            relativePath: dataItem.relativePath,
            extendProperty: dataItem.extendProperty,
            processMode: dataItem.processMode
        };
    };
    /**
     * @param {?} dataItem
     * @return {?}
     */
    TemplatemetadataComponent.prototype.onSelectedPkgChanged = /**
     * @param {?} dataItem
     * @return {?}
     */
    function (dataItem) {
        if (dataItem && dataItem.repoPkg && dataItem.repoPkg.length > 0) {
            /** @type {?} */
            var newData = this.makeNewData(dataItem);
            if (dataItem.selectedPkg) {
                newData.repoPkg.push(dataItem.selectedPkg);
            }
            else {
                newData.repoPkg.push(dataItem.repoPkg[0]);
            }
            this.selectionChanged(newData, []);
        }
    };
    /**
     * @return {?}
     */
    TemplatemetadataComponent.prototype.goSearch = /**
     * @return {?}
     */
    function () {
        if (!this.searchText) {
            this.searchText = '';
        }
        console.log('搜索');
        /** @type {?} */
        var searchTextTrim = this.searchText.trim();
        if (this.searchText && this.searchTextForDup === searchTextTrim) {
            return;
        }
        this.searchTextForDup = searchTextTrim;
        console.log(this.searchText);
        this.searchText = searchTextTrim;
        // 搜索
        console.log('search for ' + this.searchText);
        this.getData();
    };
    /**
     * @return {?}
     */
    TemplatemetadataComponent.prototype.goNext = /**
     * @return {?}
     */
    function () {
        console.log('下一页');
        this.page.pageIndex = this.page.pageIndex + 1;
        this.reload = false;
        this.getData();
    };
    /**
     * @return {?}
     */
    TemplatemetadataComponent.prototype.onConflictAvoidFlagChanged = /**
     * @return {?}
     */
    function () {
        console.log('ConflictAvoidFlag ' + this.conflictAvoidFlag);
        this.getData();
    };
    /**
     * @param {?} mdSeleted
     * @return {?}
     */
    TemplatemetadataComponent.prototype.typeSelectedChanged = /**
     * @param {?} mdSeleted
     * @return {?}
     */
    function (mdSeleted) {
        if (!mdSeleted) {
            mdSeleted = this.selectedType;
        }
        if (mdSeleted) {
            if (mdSeleted.value === this.allType.value) {
                this.postfixSelected = this.postfix;
            }
            else {
                this.postfixSelected = mdSeleted.value;
            }
            this.getData();
        }
    };
    /**
     * @param {?} source
     * @return {?}
     */
    TemplatemetadataComponent.prototype.sourceSelectedChanged = /**
     * @param {?} source
     * @return {?}
     */
    function (source) {
        if (!source) {
            source = this.selectedSource;
        }
        this.selectedSource = source;
        this.getData();
    };
    TemplatemetadataComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'app-metadatalist',
                    template: "<div class=\"templateInfo farris-main-area flex-column\">\r\n  <div class=\"templateSelectorHeader m-2\">\r\n    <div class=\"searchArea d-flex flex-row\">\r\n\r\n      <input id=\"url\" class=\"textInput flex-fill\" type=\"text\" size=\"30\" kendoTextBox [(ngModel)]=\"this.searchText\"\r\n        #url=\"ngModel\" required (ngModelChange)=\"onSearchTextChanged()\">\r\n      <button type=\"button\" class=\"btn btn-light btn-sm px-3 mx-2\"\r\n          (click)=\"goSearch()\">\u641C\u7D22</button>\r\n\r\n      <!-- <div class=\"conflictAvoidFlag\" *ngIf=\"this.scope === 2\">\r\n        <input type=\"checkbox\" class=\"form-check-input\" id=\"check1\"\r\n          [(ngModel)]=\"this.conflictAvoidFlag\" (ngModelChange)=\"this.onConflictAvoidFlagChanged()\">\r\n        <label class=\"form-check-label text-label\" for=\"check1\">\u907F\u514D\u51B2\u7A81</label>\r\n      </div> -->\r\n\r\n      <kendo-combobox class=\"mx-2\" [allowCustom]=\"allowCustom\" *ngIf=\"this.scope === 2 || this.scope === 3\" [data]=\"sourceList\" \r\n        [textField]=\"'text'\" [valueField]=\"'value'\" [(ngModel)]=\"this.selectedSource\" \r\n        (selectionChange)=\"sourceSelectedChanged($event)\" [popupSettings]=\"popupSettings\">\r\n      </kendo-combobox>\r\n\r\n      <kendo-combobox [readonly]=\"this.typeListReadonly\" [data]=\"typeList\" [(ngModel)]=\"selectedType\"\r\n        [textField]=\"'text'\" [valueField]=\"'value'\"  (selectionChange)=\"typeSelectedChanged($event)\">\r\n      </kendo-combobox>\r\n    </div>\r\n  </div>\r\n  <div class=\"templateSelector flex-fill d-flex flex-column\">\r\n    <kendo-grid class=\"flex-fill\"\r\n                [data]=\"gridData\"\r\n                [selectable]=\"selectableSetting\" \r\n                (selectionChange)=onSelectionChange($event)\r\n                [sortable]=\"true\"\r\n                [pageable]=\"false\"\r\n                [filterable]=\"false\"\r\n                [resizable]=\"true\"\r\n                [loading]=\"this.loading\"\r\n                [rowSelected]=\"rowSelected\">\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"(this.scope === 0 && this.classifier.id === 'bo') && (this.isMavenEnable === true && this.codelanguage === java)\" field=\"projectName\" title=\"\u5DE5\u7A0B\u540D\u79F0\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"150\" field=\"code\" title=\"\u7F16\u53F7\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"150\" field=\"name\" title=\"\u540D\u79F0\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column field=\"type\" [width]=\"130\" title=\"\u7C7B\u578B\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"this.scope === 1 || this.scope === 2 || this.scope === 3\" field=\"mdPkgName\"\r\n        title=\"\u5143\u6570\u636E\u5305\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"(this.scope === 2 || this.scope === 3) && (this.isMavenEnable === false || this.codelanguage === dotnet)\" field=\"nugetPkgName\" title=\"\u6240\u5C5ENuget\u5305\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"(this.scope === 2 || this.scope === 3) && (this.isMavenEnable === false || this.codelanguage === dotnet)\" field=\"nugetPkgVersion\" title=\"\u6240\u5C5ENuget\u5305\u7248\u672C\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"(this.scope === 2 || this.scope === 3) && (this.isMavenEnable === true && this.codelanguage === java)\" field=\"groupId\" title=\"\u6240\u5C5EMaven\u5305GroupId\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"(this.scope === 2 || this.scope === 3) && (this.isMavenEnable === true && this.codelanguage === java)\" field=\"artifactId\" title=\"\u6240\u5C5EMaven\u5305ArtifactId\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"(this.scope === 2 || this.scope === 3) && (this.isMavenEnable === true && this.codelanguage === java)\" field=\"mavenPkgVersion\" title=\"\u6240\u5C5EMaven\u5305\u7248\u672C\">\r\n      </kendo-grid-column>\r\n\r\n      <ng-template kendoGridDetailTemplate let-dataItem [kendoGridDetailTemplateShowIf]=\"showOnlyBeveragesDetails\">\r\n        <lib-metadatalistdetails (selectedPkgChanged)=\"this.onSelectedPkgChanged($event)\" [isMavenEnable]=\"isMavenEnable\" [mdItem]=\"dataItem\" [codelanguage]=\"codelanguage\">\r\n        </lib-metadatalistdetails>\r\n      </ng-template>\r\n\r\n    </kendo-grid>\r\n  </div>\r\n  <div *ngIf=\"this.hasNext && !this.loading\" class=\"model-footer\">\r\n    <button type=\"button\" class=\"btn btn-link btn-sm\" (click)=\"goNext()\">\u52A0\u8F7D\u66F4\u591A</button>\r\n  </div>\r\n  <div *ngIf=\"this.loading\" class=\"model-footer\">\r\n    <label class=\"text-label\">\u52A0\u8F7D\u4E2D...</label>\r\n  </div>\r\n  <div *ngIf=\"!this.hasNext && !this.loading\" class=\"model-footer\">\r\n    <label class=\"text-label\">\u5DF2\u7ECF\u5230\u5E95\u4E86</label>\r\n  </div>\r\n</div>",
                    styles: [".templateSelector{overflow-y:hidden;overflow-x:auto}.templateInfo{display:flex;overflow:hidden}.conflictAvoidFlag{margin-left:25px;align-content:baseline;margin-top:1px;margin-right:3px;min-width:65px}.scope1{min-width:800px}.scope2{min-width:1000px}.text-label{margin-top:1.3px}"]
                }] }
    ];
    /** @nocollapse */
    TemplatemetadataComponent.ctorParameters = function () { return [
        { type: MetadataService },
        { type: GSPMetadataService }
    ]; };
    TemplatemetadataComponent.propDecorators = {
        isMavenEnable: [{ type: Input }],
        technologyStack: [{ type: Output }],
        typeName: [{ type: Input }],
        filter: [{ type: Input }],
        selectedMetadataChanged: [{ type: Output }]
    };
    return TemplatemetadataComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/metadatapackagelist/metadatapackagelist.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MetadatapackagelistComponent = /** @class */ (function () {
    function MetadatapackagelistComponent(metadataService) {
        this.metadataService = metadataService;
        // tslint:disable-next-line:member-ordering
        this.state = {
            skip: 0,
            take: 9999,
            // Initial filter descriptor
            filter: {
                logic: 'and',
                filters: []
            }
        };
        // tslint:disable-next-line:member-ordering
        this.selectableSetting = { mode: 'single', checkboxOnly: false };
    }
    /**
     * @param {?} state
     * @return {?}
     */
    MetadatapackagelistComponent.prototype.dataStateChange = /**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this.state = state;
        this.gridData = process(this.data, this.state);
    };
    /**
     * @return {?}
     */
    MetadatapackagelistComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.metadataService.GetMetadataPackages().subscribe((/**
         * @param {?} packages
         * @return {?}
         */
        function (packages) {
            /** @type {?} */
            var tempdata = [];
            _this.packagesInfo = packages;
            if (_this.packagesInfo) {
                _this.packagesInfo.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    if (item.type === 'folder' || (item.type === 'file' && item.name.endsWith('mdpkg'))) {
                        tempdata.push({ 'name': item.name });
                    }
                }));
                _this.data = tempdata;
                _this.gridData = process(_this.data, _this.state);
            }
        }));
    };
    /**
     * @param {?} event
     * @return {?}
     */
    MetadatapackagelistComponent.prototype.onSelectionChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var packageName = event.selectedRows[0].dataItem.name;
        this.metadataService.clearPackage();
        this.metadataService.addPackage(packageName);
    };
    MetadatapackagelistComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'app-metadatapackagelist',
                    template: "<div class=\"templateInfo\">\r\n  <div class=\"templateSelector\">\r\n    <kendo-grid [data]=\"gridData\" \r\n                [selectable]=\"selectableSetting\" \r\n                (selectionChange)=onSelectionChange($event)\r\n                [pageSize]=\"state.take\"\r\n                [skip]=\"state.skip\"\r\n                [sort]=\"state.sort\"\r\n                [filter]=\"state.filter\"\r\n                [sortable]=\"true\"\r\n                [pageable]=\"false\"\r\n                [filterable]=\"true\"\r\n                (dataStateChange)=dataStateChange($event)>\r\n      <kendo-grid-column field=\"name\" title=\"\u5305\u540D\u79F0\" width=\"200\">\r\n      </kendo-grid-column>\r\n    </kendo-grid>\r\n  </div>\r\n</div>",
                    styles: [".templateSelector{width:580px;height:450px;border:.5px solid #ddd}.templateDescription{width:200px;height:450px;border:.5px solid #ddd}.templateInfo{display:flex}"]
                }] }
    ];
    /** @nocollapse */
    MetadatapackagelistComponent.ctorParameters = function () { return [
        { type: MetadataService }
    ]; };
    return MetadatapackagelistComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/metadatalist/metadatalistdetails/metadatalistdetails.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MetadatalistdetailsComponent = /** @class */ (function () {
    function MetadatalistdetailsComponent() {
        this.selectableSetting = { mode: 'single', checkboxOnly: false };
        this.dotnet = 'dotnet';
        this.java = 'java';
        this.detailSkip = 0;
        this.pageSize = 5;
        this.selectedPkgChanged = new EventEmitter();
        this.dataItem = [];
    }
    /**
     * @return {?}
     */
    MetadatalistdetailsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.mdItem && this.mdItem.repoPkg) {
            this.dataItem = this.mdItem.repoPkg;
            this.detailView = {
                data: this.dataItem.slice(this.detailSkip, this.detailSkip + this.pageSize),
                total: this.dataItem.length
            };
        }
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    MetadatalistdetailsComponent.prototype.detailPageChange = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var skip = _a.skip, take = _a.take;
        this.detailSkip = skip;
        this.detailView = {
            data: this.dataItem.slice(this.detailSkip, this.detailSkip + take),
            total: this.dataItem.length
        };
    };
    /**
     * @param {?} event
     * @return {?}
     */
    MetadatalistdetailsComponent.prototype.onSelectedPkgChanged = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        console.log(event);
        this.mdItem.selectedPkg = event.selectedRows[0].dataItem;
        this.selectedPkgChanged.emit(this.mdItem);
    };
    MetadatalistdetailsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-metadatalistdetails',
                    template: "<div class=\"gridDetails\">\r\n  <kendo-grid\r\n    [data]=\"detailView\"\r\n    [pageSize]=\"5\"\r\n    [skip]=\"detailSkip\"\r\n    [pageable]=\"true\"\r\n    [scrollable]=\"'none'\"\r\n    [selectable]=\"this.selectableSetting\"\r\n    (selectionChange)=onSelectedPkgChanged($event)\r\n    (pageChange)=\"detailPageChange($event)\"\r\n    [navigable]=\"true\"\r\n    kendoGridFocusable>\r\n    <kendo-grid-column field=\"name\" *ngIf=\"this.isMavenEnable === false || this.codelanguage === dotnet\" title=\"Nuget\u540D\u79F0\" width=\"120\">\r\n    </kendo-grid-column>\r\n    <kendo-grid-column field=\"version\" *ngIf=\"this.isMavenEnable === false || this.codelanguage === dotnet\" title=\"Nuget\u7248\u672C\" width=\"120\">\r\n    </kendo-grid-column>\r\n    <kendo-grid-column field=\"groupId\" *ngIf=\"this.isMavenEnable === true && this.codelanguage === java\" title=\"GroupId\" width=\"120\">\r\n    </kendo-grid-column>\r\n    <kendo-grid-column field=\"artifactId\" *ngIf=\"this.isMavenEnable === true && this.codelanguage === java\" title=\"ArtifactId\" width=\"120\">\r\n    </kendo-grid-column>\r\n    <kendo-grid-column field=\"version\" *ngIf=\"this.isMavenEnable === true && this.codelanguage === java\" title=\"Maven\u7248\u672C\" width=\"120\">\r\n    </kendo-grid-column>\r\n  </kendo-grid>\r\n</div>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    MetadatalistdetailsComponent.ctorParameters = function () { return []; };
    MetadatalistdetailsComponent.propDecorators = {
        isMavenEnable: [{ type: Input }],
        mdItem: [{ type: Input }],
        selectedPkgChanged: [{ type: Output }],
        codelanguage: [{ type: Input }]
    };
    return MetadatalistdetailsComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/selectmetadate/onclick.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var OnceClickDirective = /** @class */ (function () {
    function OnceClickDirective(renderer, // Angular 2.x导入Renderer
    element) {
        this.renderer = renderer;
        this.element = element;
        // tslint:disable-next-line:no-output-rename
        this.clickCall = new EventEmitter();
        this.duration = 1000; // 必须是数字，传入时要用绑定语法
        // 必须是数字，传入时要用绑定语法
        this.$sub = new Subject();
    }
    /**
     * @return {?}
     */
    OnceClickDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // 如此绑定事件亦可
        // this.renderer.listen(
        //   this.element.nativeElement, 'click', event => {
        //     event.preventDefault();
        //     event.stopPropagation();
        //     this.$sub.next(event);
        //   }
        // );
        this.subscription = this.$sub.pipe(throttleTime(this.duration)).subscribe((/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            _this.clickCall.emit(e);
        }));
    };
    /**
     * @param {?} event
     * @return {?}
     */
    OnceClickDirective.prototype.clickEvent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault(); // 通常是不需要冒泡的
        event.stopPropagation();
        this.$sub.next(event);
    };
    /**
     * @return {?}
     */
    OnceClickDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subscription.unsubscribe();
    };
    OnceClickDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[click.once]'
                },] }
    ];
    /** @nocollapse */
    OnceClickDirective.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    OnceClickDirective.propDecorators = {
        clickCall: [{ type: Output, args: ['click.once',] }],
        duration: [{ type: Input }],
        clickEvent: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return OnceClickDirective;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/metadata-select.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MetadataSelectModule = /** @class */ (function () {
    function MetadataSelectModule() {
    }
    MetadataSelectModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        TreeViewModule,
                        GridModule,
                        LayoutModule,
                        FarrisDialogModule,
                        DropDownsModule,
                        LoadingModule.forRoot()
                    ],
                    exports: [
                        SelectMetadataComponent
                    ],
                    providers: [MetadataService],
                    // 服务提供者
                    // tslint:disable-next-line:max-line-length
                    declarations: [
                        SelectMetadataComponent,
                        TreeclassifierComponent,
                        TemplatemetadataComponent,
                        MetadatapackageselectComponent,
                        MetadatapackagelistComponent,
                        MetadatalistdetailsComponent,
                        OnceClickDirective
                    ],
                    entryComponents: [SelectMetadataComponent, MetadatapackageselectComponent]
                },] }
    ];
    /** @nocollapse */
    MetadataSelectModule.ctorParameters = function () { return []; };
    return MetadataSelectModule;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-service.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} http
 * @param {?} url
 * @return {?}
 */
function factory(http, url) {
    return new GSPMetadataService(http, url);
}
var GSPMetadataServiceModule = /** @class */ (function () {
    function GSPMetadataServiceModule() {
    }
    /**
     * @param {?} url
     * @return {?}
     */
    GSPMetadataServiceModule.forRoot = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return {
            ngModule: GSPMetadataServiceModule,
            providers: [
                { provide: 'url', useValue: url },
                { provide: GSPMetadataService, useFactory: factory, deps: [HttpClient, 'url'] }
            ]
        };
    };
    GSPMetadataServiceModule.decorators = [
        { type: NgModule, args: [{
                    exports: [],
                    providers: []
                },] }
    ];
    return GSPMetadataServiceModule;
}());
/** @type {?} */
var globalGsp = ((/** @type {?} */ (window.top))).gsp;
if (!((/** @type {?} */ (window))).gsp) {
    ((/** @type {?} */ (window))).gsp = /*(globalGsp && globalGsp.createChild()) ||*/ new GSP(globalGsp);
}

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadatadto.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MetadataDto$1 = /** @class */ (function () {
    // tslint:disable-next-line:max-line-length
    function MetadataDto(id, nameSpace, code, name, fileName, type, bizobjectID, relativePath, extendProperty, content, extendable) {
        this.id = id;
        this.nameSpace = nameSpace;
        this.code = code;
        this.name = name;
        this.fileName = fileName;
        this.type = type;
        this.bizobjectID = bizobjectID;
        this.relativePath = relativePath;
        this.extendProperty = extendProperty;
        this.content = content;
        this.extendable = extendable;
    }
    return MetadataDto;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadataprojectdto.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MetadataProjectDto = /** @class */ (function () {
    // tslint:disable-next-line:max-line-length
    function MetadataProjectDto(packageName, projectPath, projectName, projectNameSpace, depPackageName, depPackageVersion, depPackageLocation) {
        this.packageName = packageName;
        this.projectName = projectName;
        this.projectPath = projectPath;
        this.projectNameSpace = projectNameSpace;
        this.depPackageName = depPackageName;
        this.depPackageVersion = depPackageVersion;
        this.depPackageLocation = depPackageLocation;
    }
    return MetadataProjectDto;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: gsp-lcm-metadata-selector.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { MetadataSelectModule, SelectMetadataComponent, GSPMetadataService, GSPMetadataServiceModule, MetadataDto$1 as MetadataDto, Metadata4RefDto, MetadataPackageDto, MetadataProjectDto, TemplatemetadataComponent as ɵe, MetadatalistdetailsComponent as ɵg, MetadatapackagelistComponent as ɵf, MetadatapackageselectComponent as ɵb, OnceClickDirective as ɵh, TreeclassifierComponent as ɵc, MetadataService as ɵd, factory as ɵa };
