// @farris/bef/esm5/lib/framework_session_service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/framework_session_service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: Witt
 * @Date: 2018-10-11 20:32:02
 * @Last Modified by: Witt
 * @Last Modified time: 2018-12-13 20:05:46
 */
import { Injectable, Optional } from '@angular/core';
import { SessionService } from '@ecp-caf/caf-common';
var FrameworkSessionService = /** @class */ (function () {
    function FrameworkSessionService(sessionService) {
        this.sessionService = sessionService;
    }
    /**
     * 获取用户SessionId
     */
    /**
     * 获取用户SessionId
     * @return {?}
     */
    FrameworkSessionService.prototype.getUserSessionId = /**
     * 获取用户SessionId
     * @return {?}
     */
    function () {
        return this.sessionService.getUserSessionId();
    };
    /**
     * 获取当前功能菜单的SessionId
     */
    /**
     * 获取当前功能菜单的SessionId
     * @param {?=} runtimeContext
     * @return {?}
     */
    FrameworkSessionService.prototype.getCurrentSessionId = /**
     * 获取当前功能菜单的SessionId
     * @param {?=} runtimeContext
     * @return {?}
     */
    function (runtimeContext) {
        /** @type {?} */
        var sessionId;
        /** @type {?} */
        var rsi = this.getRuntimeSessionId();
        if (rsi) {
            return rsi;
        }
        if (this.sessionService) {
            if (runtimeContext && runtimeContext.hasOwnProperty('tabId')) {
                /** @type {?} */
                var tabId = runtimeContext.tabId;
                sessionId = this.sessionService.getCurrentSeesionId(tabId);
            }
            else {
                sessionId = this.sessionService.getCurrentSeesionId();
            }
        }
        if (!sessionId) {
            sessionId = this.getUserSessionId();
        }
        return sessionId;
    };
    /**
     * @private
     * @return {?}
     */
    FrameworkSessionService.prototype.getRuntimeSessionId = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var hash = window.location.hash;
        /** @type {?} */
        var params = this.parse(hash);
        return params && params.rsi || null;
    };
    /**
     * @param {?} queryString
     * @return {?}
     */
    FrameworkSessionService.prototype.parse = /**
     * @param {?} queryString
     * @return {?}
     */
    function (queryString) {
        if (!queryString) {
            return {};
        }
        /** @type {?} */
        var hashes = queryString.slice(queryString.indexOf('?') + 1).split('&');
        return hashes.reduce((/**
         * @param {?} params
         * @param {?} hash
         * @return {?}
         */
        function (params, hash) {
            var _a;
            /** @type {?} */
            var split = hash.indexOf('=');
            /** @type {?} */
            var key = hash.slice(0, split);
            /** @type {?} */
            var val = hash.slice(split + 1);
            return Object.assign(params, (_a = {}, _a[key] = decodeURIComponent(val), _a));
        }), {});
    };
    FrameworkSessionService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FrameworkSessionService.ctorParameters = function () { return [
        { type: SessionService, decorators: [{ type: Optional }] }
    ]; };
    return FrameworkSessionService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    FrameworkSessionService.prototype.sessionService;
}
export { FrameworkSessionService };
