// @farris/bef/esm5/lib/http_service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/http_service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
// tslint:disable: no-string-literal disable: max-line-length
var HttpService = /** @class */ (function () {
    function HttpService(defaultHttpClient) {
        this.defaultHttpClient = defaultHttpClient;
        this.httpClient = defaultHttpClient;
        this.rtf = this.getService();
    }
    /**
     * 遍历获取rtf服务
     */
    /**
     * 遍历获取rtf服务
     * @private
     * @return {?}
     */
    HttpService.prototype.getService = /**
     * 遍历获取rtf服务
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var env = window;
        while (!env['gspframeworkService'] && env !== window.top && this.isSameOrigin(env)) {
            env = window.parent;
        }
        return env['gspframeworkService'] && env['gspframeworkService']['rtf'] || {};
    };
    /**
     * 是否同源
     * @param environment window
     */
    /**
     * 是否同源
     * @private
     * @param {?} environment window
     * @return {?}
     */
    HttpService.prototype.isSameOrigin = /**
     * 是否同源
     * @private
     * @param {?} environment window
     * @return {?}
     */
    function (environment) {
        /** @type {?} */
        var host = window.location.host;
        try {
            if (environment && environment.location && typeof environment.location.host !== 'undefined') {
                return environment.location.host === host;
            }
        }
        catch (e) {
            return false;
        }
        return false;
    };
    Object.defineProperty(HttpService.prototype, "params", {
        get: /**
         * @private
         * @return {?}
         */
        function () {
            if (this.rtf && this.rtf.hasOwnProperty('session') && typeof this.rtf['session']['getCommonVariable'] === 'function') {
                return this.rtf['session']['getCommonVariable']();
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HttpService.prototype, "httpService", {
        /**
         * 获取运行框架http服务
         */
        get: /**
         * 获取运行框架http服务
         * @return {?}
         */
        function () {
            if (this.rtf && this.rtf.hasOwnProperty('http') && typeof this.rtf['http']['request'] === 'function') {
                return this.rtf.http;
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} method
     * @param {?} url
     * @param {?=} options
     * @return {?}
     */
    HttpService.prototype.request = /**
     * @param {?} method
     * @param {?} url
     * @param {?=} options
     * @return {?}
     */
    function (method, url, options) {
        if (options === void 0) { options = {}; }
        // todo: 使用运行框架的httpclient导致帮助无法显示数据，暂时撤回
        // if (this.httpService && this.params && this.params.tabId) {
        //   return this.httpService.request$(method, url, options);
        // } else {
        //   return this.defaultHttpClient.request(method, url, options);
        // }
        return this.defaultHttpClient.request(method, url, options);
    };
    HttpService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    HttpService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return HttpService;
}());
export { HttpService };
if (false) {
    /** @type {?} */
    HttpService.prototype.httpClient;
    /**
     * @type {?}
     * @private
     */
    HttpService.prototype.rtf;
    /**
     * @type {?}
     * @private
     */
    HttpService.prototype.defaultHttpClient;
}
