// @farris/bef/esm5/lib/types.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/types.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: Witt
 * @Date: 2018-10-19 15:36:48
 * @Last Modified by: aalizzwell
 * @Last Modified time: 2019-07-30 17:37:02
 */
/**
 * --------------------------------------------------------------------------------
 * 变更集相关类型
 * --------------------------------------------------------------------------------
 */
/**
 * 变更类型
 */
var ChangeDetailType = /** @class */ (function () {
    function ChangeDetailType() {
    }
    /**
     * 新增
     */
    ChangeDetailType.Added = 'Added';
    /**
     * 修改
     */
    ChangeDetailType.Modify = 'Modify';
    /**
     * 删除
     */
    ChangeDetailType.Deleted = 'Deleted';
    return ChangeDetailType;
}());
export { ChangeDetailType };
if (false) {
    /**
     * 新增
     * @type {?}
     */
    ChangeDetailType.Added;
    /**
     * 修改
     * @type {?}
     */
    ChangeDetailType.Modify;
    /**
     * 删除
     * @type {?}
     */
    ChangeDetailType.Deleted;
}
/**
 * 行变更信息
 * 必须包含：
 * 1、DataId   => 主键值；
 * 2、属性名   => 新的属性值；
 * 3、子表名+s => 子表行的ChangeDetail数组
 * @record
 */
export function ChangeDetailInfo() { }
if (false) {
    /** @type {?|undefined} */
    ChangeDetailInfo.prototype.DataId;
    /* Skipping unhandled member: [key: string]: number | string | boolean | null | ChangeDetail | ChangeDetail[] | { [key: string]: any };*/
}
/**
 * 行变更详情
 * 包含：
 * 1、变更类型；
 * 2、变更信息
 * @record
 */
export function ChangeDetail() { }
if (false) {
    /** @type {?} */
    ChangeDetail.prototype.ChangeType;
    /** @type {?} */
    ChangeDetail.prototype.ChangeInfo;
}
/**
 * 请求类型
 * @record
 */
export function RequestInfo() { }
if (false) {
    /** @type {?} */
    RequestInfo.prototype.dataChange;
    /** @type {?|undefined} */
    RequestInfo.prototype.variableChange;
}
/**
 * 分页信息
 * @record
 */
export function Pagination() { }
if (false) {
    /** @type {?} */
    Pagination.prototype.pageSize;
    /** @type {?} */
    Pagination.prototype.totalCount;
    /** @type {?} */
    Pagination.prototype.pageCount;
    /** @type {?} */
    Pagination.prototype.pageIndex;
}
/**
 * 查询结果
 * @record
 */
export function QueryResult() { }
if (false) {
    /** @type {?} */
    QueryResult.prototype.result;
    /** @type {?} */
    QueryResult.prototype.pagination;
}
/**
 * 返回结果类型
 * @record
 */
export function ResponseInfo() { }
if (false) {
    /** @type {?} */
    ResponseInfo.prototype.returnValue;
    /** @type {?} */
    ResponseInfo.prototype.message;
    /** @type {?} */
    ResponseInfo.prototype.innerDataChange;
    /** @type {?} */
    ResponseInfo.prototype.innerVariableChange;
}
/**
 * 代理钩子
 * @record
 */
export function IProxyExtend() { }
if (false) {
    /**
     * @param {?} url
     * @param {?} params
     * @return {?}
     */
    IProxyExtend.prototype.extendUrl = function (url, params) { };
    /**
     * @param {?} body
     * @return {?}
     */
    IProxyExtend.prototype.extendBody = function (body) { };
    /**
     * @param {?} headers
     * @param {?=} runtimeContext
     * @return {?}
     */
    IProxyExtend.prototype.extendHeaders = function (headers, runtimeContext) { };
    /**
     * @param {?} response
     * @param {?=} ignoreChanges
     * @param {?=} options
     * @return {?}
     */
    IProxyExtend.prototype.onResponse = function (response, ignoreChanges, options) { };
    /**
     * @param {?} error
     * @param {?} selfHandError
     * @param {?} ignoreError
     * @return {?}
     */
    IProxyExtend.prototype.onError = function (error, selfHandError, ignoreError) { };
    /**
     * @param {?} response
     * @return {?}
     */
    IProxyExtend.prototype.parseHeaders = function (response) { };
}
/**
 * 带RequestInfo的body对象
 * @record
 */
export function BodyWithRequestInfo() { }
if (false) {
    /** @type {?} */
    BodyWithRequestInfo.prototype.requestInfo;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * request option格式
 * @record
 */
export function RequestOption() { }
if (false) {
    /** @type {?} */
    RequestOption.prototype.body;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * 子表分页参数
 * @record
 */
export function RetrieveParam() { }
if (false) {
    /** @type {?} */
    RetrieveParam.prototype.filters;
    /** @type {?|undefined} */
    RetrieveParam.prototype.parentIds;
}
