// @farris/mobile-command-services/esm5/lib/history-back.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/history-back.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsBridgeService } from './js-bridge-service/index';
/**
 * 帮助映射服务
 */
var /**
 * 帮助映射服务
 */
HistoryBackService = /** @class */ (function () {
    function HistoryBackService() {
    }
    /**
     * @return {?}
     */
    HistoryBackService.prototype.init = /**
     * @return {?}
     */
    function () {
        window['MOBILE_ORIGIN_BACK_ARRAY'] = window['MOBILE_ORIGIN_BACK_ARRAY'] || [];
        window['MOBILE_ORIGIN_BACK'] = {
            pushOriginGoback: /**
             * @param {?} fn
             * @return {?}
             */
            function (fn) {
                window['MOBILE_ORIGIN_BACK_ARRAY'].push(fn);
            },
            popOriginGoback: /**
             * @return {?}
             */
            function () {
                window['MOBILE_ORIGIN_BACK_ARRAY'].pop();
            },
            reflushOriginGoback: /**
             * @return {?}
             */
            function () {
                window['MOBILE_ORIGIN_BACK_ARRAY'] = [];
            }
        };
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    HistoryBackService.prototype.pushOriginGoback = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        window['MOBILE_ORIGIN_BACK_ARRAY'].push(fn);
    };
    /**
     * @return {?}
     */
    HistoryBackService.prototype.popOriginGoback = /**
     * @return {?}
     */
    function () {
        window['MOBILE_ORIGIN_BACK_ARRAY'].pop();
    };
    /**
     * @return {?}
     */
    HistoryBackService.prototype.reflushOriginGoback = /**
     * @return {?}
     */
    function () {
        window['MOBILE_ORIGIN_BACK_ARRAY'] = [];
    };
    /**
     * @return {?}
     */
    HistoryBackService.prototype.proxyBack = /**
     * @return {?}
     */
    function () {
        if (!(navigator.userAgent.toLowerCase().search('emmcloud') !== -1)) {
            return;
        }
        if (window.top.location.pathname.indexOf('mobiletaskcenter') > -1) {
            setTimeout((/**
             * @return {?}
             */
            function () {
                window.top['detailBackBtnClickHandler'] = (/**
                 * @param {?} back
                 * @return {?}
                 */
                function (back) {
                    /** @type {?} */
                    var gobackFn = window['MOBILE_ORIGIN_BACK_ARRAY'].pop();
                    if (gobackFn) {
                        gobackFn();
                        return false;
                    }
                    else {
                        /** @type {?} */
                        var key = location.pathname + location.hash.slice(0, location.hash.indexOf('?'));
                        /** @type {?} */
                        var need_go_back = sessionStorage.getItem(key);
                        if (!need_go_back) {
                            return true;
                        }
                        history.back();
                        return false;
                    }
                });
            }), 0);
        }
        else {
            /** @type {?} */
            var _close = (/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var gobackFn = window['MOBILE_ORIGIN_BACK_ARRAY'].pop();
                if (gobackFn) {
                    gobackFn();
                }
                else {
                    history.back();
                }
            });
            /** @type {?} */
            var jsBridgeService = new JsBridgeService();
            jsBridgeService.setOriginGoback(_close);
        }
    };
    return HistoryBackService;
}());
/**
 * 帮助映射服务
 */
export { HistoryBackService };
