// @farris/extend-fileupload-adapt-unifile/esm5/lib/ffilepreview-adapt-seeimg/ffilepreview-adapt-seeimg.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ffilepreview-adapt-seeimg/ffilepreview-adapt-seeimg.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, HostBinding } from '@angular/core';
import { FfilepreviewAdaptUnifileService } from './../ffilepreview-adapt-unifile.service';
var FfilepreviewAdaptSeeimgComponent = /** @class */ (function () {
    function FfilepreviewAdaptSeeimgComponent(adpSer, elementRef) {
        this.adpSer = adpSer;
        this.elementRef = elementRef;
        this.cls = true;
        // 是否启用图片的缩略图，默认为false；如果启用，图片类型会出现缩略图；不启用，根据类型显示图标
        this.enableThumbnail = false;
        // 图标的样式前缀
        this.clsPrefix = 'ffilepreview--filetype';
        // 显示缩略图中时，定义哪些后缀属于图片;
        this.supportImgSuffix = 'jpeg,jpg,gif,png,bmp';
        this.iconWidth = 38;
        // 缩略图的最大宽度
        this.maxThumbnailWidth = '100%';
        // 缩略图的最小宽度
        this.maxThumbnailHeight = '100%';
        // 处理扩展的服务器端配置，这样可以支持每个指令预览不同的目录，比如{'rootId':'default-root','formId':'formId'}
        this.extendServerConfig = null;
    }
    /**
     * @return {?}
     */
    FfilepreviewAdaptSeeimgComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * 获取图片的链接
     */
    /**
     * 获取图片的链接
     * @return {?}
     */
    FfilepreviewAdaptSeeimgComponent.prototype.imgSrc = /**
     * 获取图片的链接
     * @return {?}
     */
    function () {
        return this.adpSer.getImgSrc(this.fileInfo, this.extendServerConfig);
    };
    /**
     * 判断是否是图片
     */
    /**
     * 判断是否是图片
     * @return {?}
     */
    FfilepreviewAdaptSeeimgComponent.prototype.isImage = /**
     * 判断是否是图片
     * @return {?}
     */
    function () {
        // 没有数据
        if (!this.fileInfo) {
            return false;
        }
        // 没有名称
        /** @type {?} */
        var fileName = this.fileInfo.name;
        if (!fileName) {
            return false;
        }
        /** @type {?} */
        var lastDotIndex = fileName.lastIndexOf(".");
        /** @type {?} */
        var fileSuffix = '';
        if (lastDotIndex > -1) {
            fileSuffix = fileName.substring(lastDotIndex + 1).toLocaleLowerCase();
        }
        if (!fileSuffix) {
            return false;
        }
        /** @type {?} */
        var tImgSuffixArray = this.supportImgSuffix.split(',');
        /** @type {?} */
        var tIndex = tImgSuffixArray.findIndex((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            return item == fileSuffix;
        }));
        if (tIndex > -1) {
            return true;
        }
        return false;
    };
    /**
     * 获取文件类型样式
     */
    /**
     * 获取文件类型样式
     * @return {?}
     */
    FfilepreviewAdaptSeeimgComponent.prototype.getFileTypeClassName = /**
     * 获取文件类型样式
     * @return {?}
     */
    function () {
        /** @type {?} */
        var clsName = this.clsPrefix;
        // 如果名称为空
        if (!this.fileInfo || !this.fileInfo['name']) {
            return clsName + '-any';
        }
        /** @type {?} */
        var fileName = this.fileInfo['name'];
        /** @type {?} */
        var lastDotIndex = fileName.lastIndexOf(".");
        /** @type {?} */
        var fileSuffix = '';
        if (lastDotIndex > -1) {
            fileSuffix = fileName.substring(lastDotIndex + 1).toLocaleLowerCase();
        }
        switch (fileSuffix) {
            case 'pdf':
                clsName += '-pdf';
                break;
            case 'jpeg':
            case 'jpg':
            case 'gif':
            case 'png':
            case 'bmp':
                clsName += '-img';
                break;
            case 'ppt':
                clsName += '-ppt';
                break;
            case 'doc':
            case 'docx':
                clsName += '-doc';
                break;
            case 'xls':
            case 'xlsx':
                clsName += '-xls';
                break;
            case 'txt':
                clsName += '-txt';
                break;
            case 'zip':
                clsName += '-zip';
                break;
            default:
                clsName += '-any';
        }
        return clsName;
    };
    FfilepreviewAdaptSeeimgComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ffilepreview-adapt-seeimg',
                    template: "<div class=\"ffilepreview-seeimg--wrapper\" [ngClass]=\"{'ffilepreview-seeimg--thumbnail':enableThumbnail}\">\r\n  <ng-container *ngIf=\"enableThumbnail&&isImage();else notImage\">\r\n    <img class=\"ffilepreview-seeimg--img\" [src]=\"imgSrc()\" [ngStyle]=\"{'maxWidth':maxThumbnailWidth,'maxHeight':maxThumbnailHeight}\"/>\r\n  </ng-container>\r\n</div>\r\n<ng-template #notImage>\r\n  <span class=\"ffilepreview--filetype-icon\" [ngClass]=\"getFileTypeClassName()\" [ngStyle]=\"{'width':iconWidth+'px','height':iconWidth+'px'}\"></span>\r\n</ng-template>",
                    styles: [":host{height:100%;width:100%;position:relative}.ffilepreview-seeimg--thumbnail{top:0;bottom:0;position:absolute;right:0;left:0;display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;-webkit-box-pack:center;justify-content:center}.ffilepreview-seeimg--wrapper:hover{opacity:.8}.ffilepreview-seeimg--img{max-width:100%;max-height:100%;border-radius:4px}"]
                }] }
    ];
    /** @nocollapse */
    FfilepreviewAdaptSeeimgComponent.ctorParameters = function () { return [
        { type: FfilepreviewAdaptUnifileService },
        { type: ElementRef }
    ]; };
    FfilepreviewAdaptSeeimgComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class.ffilepreview-adapt-seeimg',] }],
        enableThumbnail: [{ type: Input }],
        clsPrefix: [{ type: Input }],
        supportImgSuffix: [{ type: Input }],
        fileInfo: [{ type: Input }],
        iconWidth: [{ type: Input }],
        maxThumbnailWidth: [{ type: Input }],
        maxThumbnailHeight: [{ type: Input }],
        extendServerConfig: [{ type: Input }]
    };
    return FfilepreviewAdaptSeeimgComponent;
}());
export { FfilepreviewAdaptSeeimgComponent };
if (false) {
    /** @type {?} */
    FfilepreviewAdaptSeeimgComponent.prototype.cls;
    /** @type {?} */
    FfilepreviewAdaptSeeimgComponent.prototype.enableThumbnail;
    /** @type {?} */
    FfilepreviewAdaptSeeimgComponent.prototype.clsPrefix;
    /** @type {?} */
    FfilepreviewAdaptSeeimgComponent.prototype.supportImgSuffix;
    /** @type {?} */
    FfilepreviewAdaptSeeimgComponent.prototype.fileInfo;
    /** @type {?} */
    FfilepreviewAdaptSeeimgComponent.prototype.iconWidth;
    /** @type {?} */
    FfilepreviewAdaptSeeimgComponent.prototype.maxThumbnailWidth;
    /** @type {?} */
    FfilepreviewAdaptSeeimgComponent.prototype.maxThumbnailHeight;
    /** @type {?} */
    FfilepreviewAdaptSeeimgComponent.prototype.extendServerConfig;
    /**
     * @type {?}
     * @private
     */
    FfilepreviewAdaptSeeimgComponent.prototype.adpSer;
    /**
     * @type {?}
     * @private
     */
    FfilepreviewAdaptSeeimgComponent.prototype.elementRef;
}
