// @edp-bif/common-api/esm5/lib/front-entity/BillTrackerSchemaTable.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/front-entity/BillTrackerSchemaTable.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BillSchemaField } from "@edp-aif/common-api";
/** @type {?} */
export var fullPathPrefix = "#";
var BillTrackerSchemaTable = /** @class */ (function () {
    function BillTrackerSchemaTable() {
    }
    /**
     * 自定义JSON序列化方法
     * @returns
     */
    /**
     * 自定义JSON序列化方法
     * @return {?}
     */
    BillTrackerSchemaTable.prototype.toJSON = /**
     * 自定义JSON序列化方法
     * @return {?}
     */
    function () {
        /** @type {?} */
        var fields = this.fields && this.fields.map((/**
         * @param {?} originF
         * @return {?}
         */
        function (originF) {
            /** @type {?} */
            var sf = new BillSchemaField();
            // sf.fullPath = originF.fullPath;
            sf.fieldPath = originF.fieldPath;
            sf.fieldName = originF.fieldName;
            sf.caption = originF.caption;
            sf.displayName = originF.displayName;
            sf.order = originF.order;
            sf.i18nCols = originF.i18nCols;
            sf.formatInfo = originF.formatInfo;
            return sf;
        }));
        /** @type {?} */
        var obj = {
            code: this.code,
            caption: this.caption,
            displayName: this.displayName,
            order: this.order,
            show: this.show,
            fields: fields,
            i18nCols: this.i18nCols,
            maxDisplayNumber: this.maxDisplayNumber,
            parentTable: this.parentTable,
        };
        return obj;
    };
    return BillTrackerSchemaTable;
}());
export { BillTrackerSchemaTable };
if (false) {
    /** @type {?} */
    BillTrackerSchemaTable.prototype.code;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.caption;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.displayName;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.order;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.show;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.fields;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.i18nCols;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.maxDisplayNumber;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.parentTable;
}
