// @farris/ui-time-picker/esm2015/lib/time-value-accessor.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { parseTime, format } from './utils/date-helper';
export class TimeValueAccessorDirective {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
    }
    /**
     * @return {?}
     */
    keyup() {
        this.changed();
    }
    /**
     * @return {?}
     */
    blur() {
        this.touched();
    }
    /**
     * @return {?}
     */
    changed() {
        if (this._onChange) {
            /** @type {?} */
            const value = parseTime(this.elementRef.nativeElement.value);
            this._onChange((/** @type {?} */ (value)));
        }
    }
    /**
     * @return {?}
     */
    touched() {
        if (this._onTouch) {
            this._onTouch();
        }
    }
    /**
     * @return {?}
     */
    setRange() {
        this.elementRef.nativeElement.focus();
        this.elementRef.nativeElement.setSelectionRange(0, this.elementRef.nativeElement.value.length);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.elementRef.nativeElement.value = format(value, this.nzTime);
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this._onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this._onTouch = fn;
    }
}
TimeValueAccessorDirective.decorators = [
    { type: Directive, args: [{
                selector: 'input[format]',
                exportAs: 'format',
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: TimeValueAccessorDirective,
                        multi: true
                    }
                ]
            },] }
];
/** @nocollapse */
TimeValueAccessorDirective.ctorParameters = () => [
    { type: ElementRef }
];
TimeValueAccessorDirective.propDecorators = {
    nzTime: [{ type: Input }],
    keyup: [{ type: HostListener, args: ['keyup',] }],
    blur: [{ type: HostListener, args: ['blur',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    TimeValueAccessorDirective.prototype._onChange;
    /**
     * @type {?}
     * @private
     */
    TimeValueAccessorDirective.prototype._onTouch;
    /** @type {?} */
    TimeValueAccessorDirective.prototype.nzTime;
    /**
     * @type {?}
     * @private
     */
    TimeValueAccessorDirective.prototype.elementRef;
}
