// @ecp-caf/caf-common/esm5/lib/services/languageservice/language.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { CookieService } from 'ngx-cookie-service';
import { CommonUtil } from '../../utilities/common.utilities';
var LanguageService = /** @class */ (function () {
    function LanguageService(cookieService) {
        this.cookieService = cookieService;
        this.languageKey = 'languageCode';
        this.curLangugaeKey = 'caf_web_language';
    }
    /**
     * @param {?} languageCode
     * @return {?}
     */
    LanguageService.prototype.setLanguageCode = /**
     * @param {?} languageCode
     * @return {?}
     */
    function (languageCode) {
        localStorage.setItem(this.languageKey, languageCode);
    };
    /**
     * @return {?}
     */
    LanguageService.prototype.getLanguageCode = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var curLancode = this.cookieService && this.cookieService.get(this.curLangugaeKey);
        if (!curLancode) {
            curLancode = CommonUtil.getCookieMapFromDoc().get(this.curLangugaeKey) ||
                localStorage.getItem(this.languageKey) ||
                'zh-CHS';
        }
        return curLancode;
    };
    LanguageService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    LanguageService.ctorParameters = function () { return [
        { type: CookieService }
    ]; };
    return LanguageService;
}());
export { LanguageService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    LanguageService.prototype.languageKey;
    /**
     * @type {?}
     * @private
     */
    LanguageService.prototype.curLangugaeKey;
    /**
     * @type {?}
     * @private
     */
    LanguageService.prototype.cookieService;
}
