// @gsp-aif/common-api/esm2015/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**Aif Common Entity (AIF公共实体类) */
/**调用类型 */
export { AifCallType } from './lib/aif-call-type';
/**业务动作类型 */
export { AifBizActionType } from './lib/aif-biz-action-type';
/**业务动作实体类 */
export { AifBizActionEntity } from './lib/entity/impl/aif-biz-action-entity';
/**业务场景图实体类 */
export { AifBizFlowchartEntity } from './lib/entity/impl/aif-biz-flowchart-entity';
/**业务事件实体类 */
export { AifBizEventEntity } from './lib/entity/impl/aif-biz-event-entity';
/**生单规则实体类 */
export { AifCreationRuleEntity } from './lib/entity/impl/aif-creation-rule-entity';
/**回写规则实体类 */
export { AifWritebackRuleEntity } from './lib/entity/impl/aif-writeback-rule-entity';
/**生单流程实体类 */
export { AifCreationProcessEntity } from './lib/entity/impl/aif-creation-process-entity';
/**回写流程实体类 */
export { AifWritebackProcessEntity } from './lib/entity/impl/aif-writeback-process-entity';
