// @farris/ide-devkit/esm5/lib/ideframework/gsp/ide/messager.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var Messager = /** @class */ (function () {
    function Messager() {
    }
    /**
     * @param {?} message
     * @param {?=} callback
     * @return {?}
     */
    Messager.prototype.info = /**
     * @param {?} message
     * @param {?=} callback
     * @return {?}
     */
    function (message, callback) {
    };
    /**
     * @param {?} message
     * @param {?} okCallback
     * @param {?=} cancelCallback
     * @return {?}
     */
    Messager.prototype.question = /**
     * @param {?} message
     * @param {?} okCallback
     * @param {?=} cancelCallback
     * @return {?}
     */
    function (message, okCallback, cancelCallback) {
    };
    /**
     * @param {?} message
     * @return {?}
     */
    Messager.prototype.confirm = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
        return null;
    };
    /**
     * @param {?} message
     * @return {?}
     */
    Messager.prototype.error = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
    };
    /**
     * @param {?} message
     * @return {?}
     */
    Messager.prototype.warning = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
    };
    return Messager;
}());
export { Messager };
