// @qdp/common/esm2015/lib/service/utils/json.util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class JSONUtils {
    /**
     * @param {?} object
     * @return {?}
     */
    static isObj(object) {
        return object && typeof (object) == 'object' && Object.prototype.toString.call(object).toLowerCase() == "[object object]";
    }
    /**
     * @param {?} object
     * @return {?}
     */
    static isArray(object) {
        return object && typeof (object) == 'object' && object.constructor == Array;
    }
    /**
     * @param {?} object
     * @return {?}
     */
    static getLength(object) {
        /** @type {?} */
        var count = 0;
        for (var i in object)
            count++;
        return count;
    }
    /**
     * @param {?} objA
     * @param {?} objB
     * @return {?}
     */
    static Compare(objA, objB) {
        if (!JSONUtils.isObj(objA) || !JSONUtils.isObj(objB))
            return false; //判断类型是否正确
        if (JSONUtils.getLength(objA) != JSONUtils.getLength(objB))
            return false; //判断长度是否一致
        return JSONUtils.CompareObj(objA, objB, true); //默认为true
    }
    /**
     * @private
     * @param {?} objA
     * @param {?} objB
     * @param {?} flag
     * @return {?}
     */
    static CompareObj(objA, objB, flag) {
        for (var key in objA) {
            if (!flag) //跳出整个循环
                break;
            if (!objB.hasOwnProperty(key)) {
                flag = false;
                break;
            }
            if (!JSONUtils.isArray(objA[key])) { //子级不是数组时,比较属性值
                if (objB[key] != objA[key]) {
                    flag = false;
                    break;
                }
            }
            else {
                if (!JSONUtils.isArray(objB[key])) {
                    flag = false;
                    break;
                }
                /** @type {?} */
                var oA = objA[key];
                /** @type {?} */
                var oB = objB[key];
                if (oA.length != oB.length) {
                    flag = false;
                    break;
                }
                for (var k in oA) {
                    if (!flag) //这里跳出循环是为了不让递归继续
                        break;
                    flag = JSONUtils.CompareObj(oA[k], oB[k], flag);
                }
            }
        }
        return flag;
    }
}
