// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/param-serialzier/base-param-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/param-serialzier/base-param-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DefaultParameterSerializer } from '@ecp-caf/common-structure';
import { BaseParameter } from '../../extension/parameter/base-param';
var BaseParameterSerializer = /** @class */ (function () {
    function BaseParameterSerializer() {
    }
    /**
     * @param {?} param
     * @param {?} ctxt
     * @return {?}
     */
    BaseParameterSerializer.prototype.serialize = /**
     * @param {?} param
     * @param {?} ctxt
     * @return {?}
     */
    function (param, ctxt) {
        /** @type {?} */
        var jobj = new Object();
        if (!param) {
            return jobj;
        }
        /** @type {?} */
        var baseParam = (/** @type {?} */ (param));
        /** @type {?} */
        var serializer = new DefaultParameterSerializer();
        jobj = serializer.serialize(baseParam, ctxt);
        jobj['id'] = baseParam.id;
        jobj['index'] = baseParam.index;
        jobj['visible'] = baseParam.visible;
        jobj['readonly'] = baseParam.readonly;
        jobj['groupId'] = baseParam.groupId;
        jobj['groupCode'] = baseParam.groupCode;
        jobj['groupName'] = baseParam.groupName;
        jobj['parameterTypeDesc'] = baseParam.parameterTypeDesc;
        jobj['kind'] = baseParam.kind;
        return jobj;
    };
    /**
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    BaseParameterSerializer.prototype.deserialize = /**
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    function (obj, ctxt) {
        if (!obj) {
            return null;
        }
        /** @type {?} */
        var param = new BaseParameter();
        /** @type {?} */
        var serializer = new DefaultParameterSerializer();
        param = (/** @type {?} */ (serializer.deserialize(obj, ctxt)));
        param.id = obj['id'];
        param.index = (/** @type {?} */ (obj['index']));
        param.visible = (/** @type {?} */ (obj['visible']));
        param.readonly = (/** @type {?} */ (obj['readonly']));
        param.groupId = obj['groupId'];
        param.groupCode = obj['groupCode'];
        param.groupName = obj['groupName'];
        param.parameterTypeDesc = obj['parameterTypeDesc'];
        return param;
    };
    return BaseParameterSerializer;
}());
export { BaseParameterSerializer };
