// @gsp-wf/wf-process-model/esm5/lib/model/extension/tool/Component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { Tool } from "./Tool";
var Component = /** @class */ (function (_super) {
    tslib_1.__extends(Component, _super);
    function Component(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = Component.Clr_Type_ID;
        return _this;
    }
    /**
     * @return {?}
     */
    Component.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new Component(this.Model);
    };
    //Json 序列化
    //Json 序列化
    /**
     * @return {?}
     */
    Component.prototype.ConvertToJsonObject = 
    //Json 序列化
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "MetadataId", this.MetadataId);
        BpmnModelHelper.SetAttributeValue(obj, "ComponentType", this.ComponentType);
        BpmnModelHelper.SetAttributeValue(obj, "MethodCode", this.MethodCode);
        BpmnModelHelper.SetAttributeValue(obj, "ScopeType", this.ScopeType);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    Component.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.MetadataId = obj.MetadataId;
        this.ComponentType = obj.ComponentType;
        this.MethodCode = obj.MethodCode;
        this.ScopeType = obj.ScopeType;
    };
    /**
     * @return {?}
     */
    Component.prototype.clone = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        /** @type {?} */
        var component = new Component(this.Model);
        component.Id = BpmnModelHelper.GenerateElementId();
        component.Code = this.Code;
        component.Name = this.Name;
        component.Description = this.Description;
        component.Execution = this.Execution;
        component.AsyncCallBackType = this.AsyncCallBackType;
        component.Time = this.Time;
        component.isForCompensation = this.isForCompensation;
        if (this.ActualParameters && this.ActualParameters.length > 0) {
            try {
                for (var _b = tslib_1.__values(this.ActualParameters), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var p = _c.value;
                    /** @type {?} */
                    var parameter = p.clone();
                    component.ActualParameters.push(parameter);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        component.MetadataId = this.MetadataId;
        component.ComponentType = this.ComponentType;
        component.MethodCode = this.MethodCode;
        component.ScopeType = this.ScopeType;
        return component;
    };
    Component.Clr_Type_ID = "wf.Component";
    return Component;
}(Tool));
export { Component };
if (false) {
    /** @type {?} */
    Component.Clr_Type_ID;
    /** @type {?} */
    Component.prototype.ClrTypeID;
    /** @type {?} */
    Component.prototype.MetadataId;
    /** @type {?} */
    Component.prototype.ComponentType;
    /** @type {?} */
    Component.prototype.MethodCode;
    /** @type {?} */
    Component.prototype.ScopeType;
}
BpmnModelHelper.RegisterElementType(new Component(new BpmnModel()));
