// @gsp-wf/wf-process-model/esm5/lib/model/extension/tool/Tool.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { AsyncCallBackType } from "../../misc/AsyncCallBackType";
import { ExecutionType } from "../../misc/ExecutionType";
import { ActualParameter } from "../../parameter/WFActualParameter";
import { WFExtensionElement } from "../WFExtensionElement";
import { ToolTime } from "./ToolTime";
/**
 * @abstract
 */
var Tool = /** @class */ (function (_super) {
    tslib_1.__extends(Tool, _super);
    function Tool(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = Tool.Clr_Type_ID;
        _this.ActualParameters = new Array();
        _this.Execution = ExecutionType.SYNCHRONOUS;
        _this.AsyncCallBackType = AsyncCallBackType.NO;
        _this.Time = ToolTime.UNDEFINITION;
        _this.isForCompensation = false;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    Tool.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    Tool.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        /** @type {?} */
        var actualParametersElements = [];
        if (this.ActualParameters !== null && this.ActualParameters.length > 0) {
            this.ActualParameters.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                actualParametersElements.push(item.ConvertToJsonObject());
            }));
        }
        BpmnModelHelper.SetAttributeValue(obj, "ActualParameters", actualParametersElements);
        BpmnModelHelper.SetAttributeValue(obj, 'Execution', this.Execution);
        BpmnModelHelper.SetAttributeValue(obj, 'Time', this.Time);
        BpmnModelHelper.SetAttributeValue(obj, 'IsForCompensation', this.isForCompensation);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    Tool.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        _super.prototype.LoadFromJsonObject.call(this, obj);
        /** @type {?} */
        var actualParameters = (/** @type {?} */ (obj.ActualParameters));
        if (actualParameters !== null && actualParameters.length > 0) {
            actualParameters.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                /** @type {?} */
                var actualParameter = new ActualParameter(_this.Model);
                actualParameter.LoadFromJsonObject(item);
                _this.ActualParameters.push(actualParameter);
            }));
        }
        this.Execution = (/** @type {?} */ (obj.Execution));
        this.Time = (/** @type {?} */ (obj.Time));
        this.isForCompensation = obj.IsForCompensation;
    };
    Tool.Clr_Type_ID = "wf.Tool";
    return Tool;
}(WFExtensionElement));
export { Tool };
if (false) {
    /** @type {?} */
    Tool.Clr_Type_ID;
    /** @type {?} */
    Tool.prototype.ClrTypeID;
    /** @type {?} */
    Tool.prototype.ActualParameters;
    /** @type {?} */
    Tool.prototype.Execution;
    /** @type {?} */
    Tool.prototype.AsyncCallBackType;
    /** @type {?} */
    Tool.prototype.Time;
    /** @type {?} */
    Tool.prototype.isForCompensation;
}
