// @gsp-lcm/dbo-selectrt/esm2015/lib/dbo-selectrt.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ComponentFactoryResolver, Injector, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DboSelectRTComponent } from './dbo-selectrt/dbo-selectrt.component';
import { GridModule } from '@progress/kendo-angular-grid';
import { FormsModule } from '@angular/forms';
import { PackageModule } from '@farris/ide-devkit';
import { DboHelpRtService } from './dbo-help-rt.service';
import { HttpClientModule } from '@angular/Common/http';
export class DboSelectRTModule extends PackageModule {
    /**
     * @param {?} cfr
     * @param {?} injector
     */
    constructor(cfr, injector) {
        super();
        this.cfr = cfr;
        this.injector = injector;
    }
}
DboSelectRTModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    GridModule,
                    FormsModule,
                    HttpClientModule
                ],
                providers: [DboHelpRtService],
                declarations: [DboSelectRTComponent],
                entryComponents: [DboSelectRTComponent]
            },] }
];
/** @nocollapse */
DboSelectRTModule.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: Injector }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    DboSelectRTModule.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    DboSelectRTModule.prototype.injector;
}
