// @gsp-lcm/dbo-selectrt/esm2015/lib/dbo-help-rt.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpHeaders, HttpClient } from '@angular/Common/http';
export class DboHelpRtService {
    /**
     * @param {?} rest
     */
    constructor(rest) {
        this.rest = rest;
        this.databaseurlRT = '/api/runtime/sys/v1.0/database-objects';
    }
    /**
     * 数据库对象选择帮助界面（运行时）
     * @return {?}
     */
    getAllDatabaseObjectRT() {
        /** @type {?} */
        var headers = new HttpHeaders();
        headers.append('Content-Type', 'application/json');
        headers.append('Authorization', 'as');
        headers.append('Access-Control-Allow-Origin', '*');
        headers.append('sessionId', this.sessionId);
        /** @type {?} */
        var options = {
            params: {},
            headers: headers
        };
        return this.rest.get(this.databaseurlRT, options).pipe();
    }
    /**
     * 左侧业务对象导航树右侧dbo列表帮助界面
     * @param {?} boID
     * @return {?}
     */
    getAllDatabaseObjectByBOID(boID) {
        /** @type {?} */
        const headers = new HttpHeaders();
        headers.append('Content-Type', 'application/json');
        headers.append('Authorization', 'as');
        headers.append('Access-Control-Allow-Origin', '*');
        headers.append('sessionId', this.sessionId);
        /** @type {?} */
        const options = {
            params: {},
            headers: headers
        };
        /** @type {?} */
        const url = this.databaseurlRT + '/bo/' + boID;
        return this.rest.get(url, options).pipe();
    }
}
DboHelpRtService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DboHelpRtService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    DboHelpRtService.prototype.databaseurlRT;
    /** @type {?} */
    DboHelpRtService.prototype.sessionId;
    /**
     * @type {?}
     * @private
     */
    DboHelpRtService.prototype.rest;
}
