// @gsp-svc/file-viewer/esm2015/lib/image-viewer/image-viewer/image-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
export class ImageViewerComponent {
    constructor() {
        this.index = 0;
        this.style = { transform: '', msTransform: '', oTransform: '', webkitTransform: '' };
        this.scale = 1;
        this.rotation = 0;
        this.translateX = 0;
        this.translateY = 0;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    zoomIn() {
        this.scale < 4 ? this.scale += 0.1 : 4;
        this.updateStyle();
    }
    /**
     * @return {?}
     */
    zoomOut() {
        this.scale > 0.3 ? this.scale -= 0.1 : 0.3;
        this.scale = Number(this.scale.toFixed(1));
        this.updateStyle();
    }
    /**
     * @return {?}
     */
    zoomReset() {
        this.scale = 1;
        this.updateStyle();
    }
    /**
     * @param {?} evt
     * @return {?}
     */
    scrollZoom(evt) {
        evt.deltaY > 0 ? this.zoomOut() : this.zoomIn();
        return false;
    }
    /**
     * @return {?}
     */
    rotateClockwise() {
        this.rotation += 90;
        this.updateStyle();
    }
    /**
     * @return {?}
     */
    rotateCounterClockwise() {
        this.rotation -= 90;
        this.updateStyle();
    }
    /**
     * @param {?} evt
     * @return {?}
     */
    onDragStart(evt) {
        this.prevX = evt.clientX;
        this.prevY = evt.clientY;
    }
    /**
     * @param {?} evt
     * @return {?}
     */
    onDragOver(evt) {
        this.translateX += (evt.clientX - this.prevX);
        this.translateY += (evt.clientY - this.prevY);
        this.prevX = evt.clientX;
        this.prevY = evt.clientY;
        this.updateStyle();
    }
    // @HostListener('window:keyup.ArrowLeft')
    // preFile() {
    //   if (this.index > 0) {
    //     this.index -= 1;
    //     this.reset();
    //   }
    // }
    // @HostListener('window:keyup.ArrowRight')
    // nextFile() {
    //   if (this.index < this.src.length - 1) {
    //     this.index += 1;
    //     this.reset();
    //   }
    // }
    /**
     * @param {?} url
     * @return {?}
     */
    imageChange(url) {
        this.imageUrl = url;
        this.reset();
    }
    /**
     * @return {?}
     */
    reset() {
        this.scale = 1;
        this.rotation = 0;
        this.translateX = 0;
        this.translateY = 0;
        this.updateStyle();
    }
    /**
     * @return {?}
     */
    updateStyle() {
        this.style.transform = `translate(${this.translateX}px, ${this.translateY}px) rotate(${this.rotation}deg) scale(${this.scale})`;
        this.style.msTransform = this.style.transform;
        this.style.webkitTransform = this.style.transform;
        this.style.oTransform = this.style.transform;
    }
}
ImageViewerComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-image-viewer',
                template: "<div (wheel)=\"scrollZoom($event)\" (dragover)=\"onDragOver($event)\" (dragstart)=\"onDragStart($event)\"  >\n  <img [src]='imageUrl' id=\"image\" [ngStyle]=\"style\">\n  <!-- <div>\n      <button id=\"zoomin\" kendoButton [icon]=\"'zoom-in'\" (click)=\"zoomIn()\"></button>\n      <button id=\"zoomout\" kendoButton [icon]=\"'zoom-out'\" (click)=\"zoomOut()\"></button>\n      <button id=\"clockwise\" kendoButton [icon]=\"'reload'\" (click)=\"rotateClockwise()\"></button>\n      <button id=\"counterclockwise\" kendoButton [icon]=\"'reset'\" (click)=\"rotateCounterClockwise()\"></button>\n  </div> -->\n  <!--div id=\"picchange\">\n      <button *ngIf=\"src.length>1\" id=\"prepic\" kendoButton [icon]=\"'arrow-chevron-left'\" (click)=\"prePicture()\"\n          [disabled]=\"index == 0\"></button>\n\n      <button *ngIf=\"src.length>1\" id=\"nextpic\" kendoButton [icon]=\"'arrow-chevron-right'\" (click)=\"nextPicture()\"\n          [disabled]=\"index == src.length - 1\"></button>\n  </div-->\n</div>\n",
                styles: ["#image{text-align:center;margin:auto;top:0;left:0;bottom:0;right:0;position:absolute;max-height:100%;max-width:100%;overflow:hidden}#zoomin{width:40px;height:40px;border-radius:20px;position:absolute;right:30px;bottom:285px;transform:scale(1.5);-ms-transform:scale(1.5);-moz-transform:scale(1.5);-webkit-transform:scale(1.5);-o-transform:scale(1.5)}#zoomout{width:40px;height:40px;border-radius:20px;position:absolute;right:30px;bottom:205px;transform:scale(1.5);-ms-transform:scale(1.5);-moz-transform:scale(1.5);-webkit-transform:scale(1.5);-o-transform:scale(1.5)}#clockwise{width:40px;height:40px;border-radius:20px;position:absolute;right:30px;bottom:125px;transform:scale(1.5) rotate(45deg);-ms-transform:scale(1.5) rotate(45deg);-moz-transform:scale(1.5) rotate(45deg);-webkit-transform:scale(1.5) rotate(45deg);-o-transform:scale(1.5) rotate(45deg)}#counterclockwise{width:40px;height:40px;border-radius:20px;position:absolute;right:30px;bottom:45px;transform:scale(1.5) rotate(-45deg);-ms-transform:scale(1.5) rotate(-45deg);-moz-transform:scale(1.5) rotate(-45deg);-webkit-transform:scale(1.5) rotate(-45deg);-o-transform:scale(1.5) rotate(-45deg)}#prepic{width:40px;height:40px;border-radius:20px;position:absolute;bottom:25px;right:50px;transform:scale(1.5);-ms-transform:scale(1.5);-moz-transform:scale(1.5);-webkit-transform:scale(1.5);-o-transform:scale(1.5)}#nextpic{width:40px;height:40px;border-radius:20px;position:relative;bottom:25px;left:50px;transform:scale(1.5);-ms-transform:scale(1.5);-moz-transform:scale(1.5);-webkit-transform:scale(1.5);-o-transform:scale(1.5)}#picchange{position:fixed;bottom:0;left:50%}"]
            }] }
];
/** @nocollapse */
ImageViewerComponent.ctorParameters = () => [];
ImageViewerComponent.propDecorators = {
    imageUrl: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    ImageViewerComponent.prototype.imageUrl;
    /** @type {?} */
    ImageViewerComponent.prototype.index;
    /** @type {?} */
    ImageViewerComponent.prototype.style;
    /** @type {?} */
    ImageViewerComponent.prototype.scale;
    /** @type {?} */
    ImageViewerComponent.prototype.rotation;
    /** @type {?} */
    ImageViewerComponent.prototype.translateX;
    /** @type {?} */
    ImageViewerComponent.prototype.translateY;
    /** @type {?} */
    ImageViewerComponent.prototype.prevX;
    /** @type {?} */
    ImageViewerComponent.prototype.prevY;
}
