// @gsp-wf/task-action-impl/esm5/lib/throttleTime/throttleClickDirective.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, HostListener, Output } from '@angular/core';
import { Subject } from 'rxjs';
import { throttleTime } from 'rxjs/operators';
/**
 * 点击节流器：避免连续点击
 */
var ThrottleClickDirective = /** @class */ (function () {
    function ThrottleClickDirective() {
        // 一定的时间内只执行第一个事件
        this.THROTTLE_TIME = 2000;
        this.subject = new Subject();
        // 事件方法
        this.throttleClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ThrottleClickDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var result = this.subject.pipe(throttleTime(this.THROTTLE_TIME));
        this.click = result.subscribe(function (e) {
            _this.throttleClick.emit(e);
        });
    };
    /**
     * @return {?}
     */
    ThrottleClickDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.click.unsubscribe();
    };
    /**
     * @param {?} evt
     * @return {?}
     */
    ThrottleClickDirective.prototype.onClick = /**
     * @param {?} evt
     * @return {?}
     */
    function (evt) {
        this.subject.next(evt);
    };
    ThrottleClickDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[directive-throttle]'
                },] }
    ];
    /** @nocollapse */
    ThrottleClickDirective.ctorParameters = function () { return []; };
    ThrottleClickDirective.propDecorators = {
        throttleClick: [{ type: Output }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return ThrottleClickDirective;
}());
export { ThrottleClickDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThrottleClickDirective.prototype.THROTTLE_TIME;
    /**
     * @type {?}
     * @private
     */
    ThrottleClickDirective.prototype.subject;
    /**
     * @type {?}
     * @private
     */
    ThrottleClickDirective.prototype.click;
    /** @type {?} */
    ThrottleClickDirective.prototype.throttleClick;
}
