// @gspwidget/util/esm5/lib/rtf-util.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
/** @type {?} */
var WINDOW_RTF_KEY = 'gspframeworkService';
/**
 * @record
 */
export function RtfTabChangeEvent() { }
if (false) {
    /** @type {?} */
    RtfTabChangeEvent.prototype.appEntrance;
    /** @type {?} */
    RtfTabChangeEvent.prototype.appId;
    /** @type {?} */
    RtfTabChangeEvent.prototype.appType;
    /** @type {?} */
    RtfTabChangeEvent.prototype.funcId;
    /** @type {?} */
    RtfTabChangeEvent.prototype.tabId;
    /** @type {?} */
    RtfTabChangeEvent.prototype.tabName;
}
var RtfUtilService = /** @class */ (function () {
    function RtfUtilService(translate) {
        this.translate = translate;
    }
    // TODO: queryParamArray 是个 array 就行
    // TODO: queryParamArray 是个 array 就行
    /**
     * @param {?} funcId
     * @param {?=} queryParamArray
     * @param {?=} tabId
     * @param {?=} isReload
     * @return {?}
     */
    RtfUtilService.prototype.openMenu = 
    // TODO: queryParamArray 是个 array 就行
    /**
     * @param {?} funcId
     * @param {?=} queryParamArray
     * @param {?=} tabId
     * @param {?=} isReload
     * @return {?}
     */
    function (funcId, queryParamArray, tabId, isReload) {
        ((/** @type {?} */ (window))).gspframeworkService.rtf.func.openMenu((/** @type {?} */ ({
            appType: 'menu',
            funcId: funcId, tabId: tabId, isReload: isReload,
            appId: '', appEntrance: '',
            queryStringParams: new Map(queryParamArray)
        })));
    };
    /**
     * @param {?} funcId
     * @param {?=} tabId
     * @return {?}
     */
    RtfUtilService.prototype.closeMenu = /**
     * @param {?} funcId
     * @param {?=} tabId
     * @return {?}
     */
    function (funcId, tabId) {
        ((/** @type {?} */ (window))).gspframeworkService.rtf.func.close((/** @type {?} */ ({
            appType: 'menu',
            funcId: funcId, tabId: tabId,
            appId: '', appEntrance: ''
        })));
    };
    /**
     * @param {?} appId
     * @return {?}
     */
    RtfUtilService.prototype.openApp = /**
     * @param {?} appId
     * @return {?}
     */
    function (appId) {
        ((/** @type {?} */ (window))).gspframeworkService.rtf.func.openMenu((/** @type {?} */ ({
            appType: 'app',
            funcId: '',
            appId: appId,
            appEntrance: '',
        })));
    };
    /**
     * @return {?}
     */
    RtfUtilService.prototype.openUserInfoMenu = /**
     * @return {?}
     */
    function () {
        var rtf = window[WINDOW_RTF_KEY].rtf;
        /** @type {?} */
        var frmEventSvc = rtf.frmEvent;
        /** @type {?} */
        var sessionSvc = rtf.adapter.adatperSerice().session;
        this.translate.get('userInfo').subscribe((/**
         * @param {?} txt
         * @return {?}
         */
        function (txt) {
            frmEventSvc.eventFire("farrisapp-click", {
                active: false,
                code: 'userInfo',
                src: '/platform/runtime/sys/web/userinfo/index.html',
                sessionid: sessionSvc.getUserSessionId(),
                id: 'userinfo',
                appType: '',
                FuncName: txt
            });
        }));
    };
    /**
     * @param {?} route
     * @return {?}
     */
    RtfUtilService.prototype.closeCurRtfTab = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        var _a = route.snapshot.queryParams, tabId = _a.tabId, funcId = _a.funcId;
        this.closeMenu(funcId, tabId);
    };
    /**
     * @param {?} route
     * @param {?=} condition
     * @return {?}
     */
    RtfUtilService.prototype.onRtfTabChangeToMe = /**
     * @param {?} route
     * @param {?=} condition
     * @return {?}
     */
    function (route, condition) {
        /** @type {?} */
        var onRtfTabChangeToMe$ = new EventEmitter();
        var curTabId = route.snapshot.queryParams.tabId;
        onRtfTabChange().subscribe((/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            if (!condition)
                condition = (/**
                 * @param {?} pToTabId
                 * @param {?} pCurTabId
                 * @return {?}
                 */
                function (pToTabId, pCurTabId) { return pToTabId === pCurTabId; });
            if (condition(e.tabId, curTabId)) {
                onRtfTabChangeToMe$.emit(e);
            }
        }));
        return onRtfTabChangeToMe$;
    };
    RtfUtilService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    RtfUtilService.ctorParameters = function () { return [
        { type: TranslateService }
    ]; };
    return RtfUtilService;
}());
export { RtfUtilService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    RtfUtilService.prototype.translate;
}
/** @type {?} */
var onRtfTabChange$ = new EventEmitter();
/** @type {?} */
var funcSwitchEventRegistered = false
/** 用函数是为了延迟注册事件，否则 rtf service 可能还没挂在 window 上 */
;
/**
 * 用函数是为了延迟注册事件，否则 rtf service 可能还没挂在 window 上
 * @return {?}
 */
export function onRtfTabChange() {
    if (!funcSwitchEventRegistered && window[WINDOW_RTF_KEY]) {
        funcSwitchEventRegistered = true;
        window[WINDOW_RTF_KEY].rtf.frmEvent.eventListener('funcSwitchEvent', (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            onRtfTabChange$.emit(e);
        }));
    }
    return onRtfTabChange$;
}
