// @farris/ui-datatable/esm5/lib/utils/drag.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, NgZone, Renderer2 } from '@angular/core';
import { DataTableComponent } from '../datatable.component';
var DragColumnDirective = /** @class */ (function () {
    function DragColumnDirective(dt, ngzone, el, render) {
        this.dt = dt;
        this.ngzone = ngzone;
        this.el = el;
        this.render = render;
    }
    /**
     * @return {?}
     */
    DragColumnDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        // if (this.isEnable()) {
        //     this.ngzone.runOutsideAngular(() => {
        //         // this.resizer = document.createElement('span');
        //         // this.resizer.className = 'column-resizer';
        //         // this.el.nativeElement.appendChild(this.resizer);
        //         // this.render.addClass(this.el.nativeElement, 'resizable-column');
        //         this.resizer = this.el.nativeElement.querySelector('span.column-resizer');
        //         if (this.resizer) {
        //             this.resizerMouseDownListener = this.onMouseDown.bind(this);
        //             this.resizer.addEventListener('mousedown', this.resizerMouseDownListener);
        //         }
        //     });
        // }
    };
    /**
     * @return {?}
     */
    DragColumnDirective.prototype.isEnable = /**
     * @return {?}
     */
    function () {
        return this.dt.resizableColumns && !this.dt.hasFixed;
    };
    /**
     * @return {?}
     */
    DragColumnDirective.prototype.bindDocumentEvents = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.ngzone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            _this.documentMouseMoveListener = _this.onMouseMove.bind(_this);
            document.addEventListener('mousemove', _this.documentMouseMoveListener);
            _this.documentMouseUpListener = _this.onMouseUp.bind(_this);
            document.addEventListener('mouseup', _this.documentMouseUpListener);
        }));
    };
    /**
     * @return {?}
     */
    DragColumnDirective.prototype.unbindDocumentEvents = /**
     * @return {?}
     */
    function () {
        if (this.documentMouseMoveListener) {
            document.removeEventListener('mousemove', this.documentMouseMoveListener);
            this.documentMouseMoveListener = null;
        }
        if (this.documentMouseUpListener) {
            document.removeEventListener('mouseup', this.documentMouseUpListener);
            this.documentMouseUpListener = null;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DragColumnDirective.prototype.onMouseDown = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.dt.beginDrag(event);
        this.bindDocumentEvents();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DragColumnDirective.prototype.onMouseMove = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.dt.moveDrag(event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DragColumnDirective.prototype.onMouseUp = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.dt.stopDrag(event, this.el.nativeElement);
        this.unbindDocumentEvents();
    };
    /**
     * @return {?}
     */
    DragColumnDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        // if (this.resizerMouseDownListener && this.resizer) {
        //     this.resizer.removeEventListener('mousedown', this.resizerMouseDownListener);
        // }
        this.unbindDocumentEvents();
    };
    DragColumnDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[drag-column]',
                    exportAs: 'resizeRef'
                },] }
    ];
    /** @nocollapse */
    DragColumnDirective.ctorParameters = function () { return [
        { type: DataTableComponent },
        { type: NgZone },
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    return DragColumnDirective;
}());
export { DragColumnDirective };
if (false) {
    /** @type {?} */
    DragColumnDirective.prototype.resizer;
    /** @type {?} */
    DragColumnDirective.prototype.resizerMouseDownListener;
    /** @type {?} */
    DragColumnDirective.prototype.documentMouseMoveListener;
    /** @type {?} */
    DragColumnDirective.prototype.documentMouseUpListener;
    /** @type {?} */
    DragColumnDirective.prototype.dt;
    /** @type {?} */
    DragColumnDirective.prototype.ngzone;
    /** @type {?} */
    DragColumnDirective.prototype.el;
    /** @type {?} */
    DragColumnDirective.prototype.render;
}
