// @farris/ide-devkit/esm2015/lib/ideframework/package/package.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PackageView } from './package-view';
export class Package {
    /**
     * @param {?} gsp
     */
    constructor(gsp) {
        this.gsp = gsp;
        // leftPanel: any;
        // modelPanel: any;
        this.subscriptions = [];
        this.packageId = gsp.util.newGuid();
    }
    /**
     * @return {?}
     */
    get id() {
        return this.packageId;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set id(value) {
        this.packageId = value;
    }
    /**
     * @return {?}
     */
    get view() {
        return this.packageView;
    }
    /**
     * @param {?} state
     * @return {?}
     */
    activate(state) {
        this.packageView = new PackageView(state.sagipackageViewState);
        // /*this.leftPanel = */this.ws.addLeftPanel({
        //   title: 'test',
        //   item: this.packageView.getComponent(),
        //   visible: false
        // });
    }
    /**
     * @return {?}
     */
    initialize() { }
    /**
     * @return {?}
     */
    deactivate() {
        this.packageView.destroy();
    }
    /**
     * @return {?}
     */
    serialize() {
        return {
            sagipackageViewState: this.packageView.serialize()
        };
    }
    /**
     * @return {?}
     */
    toggle() {
        // return (
        //   this.modelPanel.isVisible() ?
        //     this.modelPanel.hide() :
        //     this.modelPanel.show()
        // );
    }
    /**
     * @return {?}
     */
    getComponent() {
        return null;
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    Package.prototype.packageView;
    /**
     * @type {?}
     * @protected
     */
    Package.prototype.subscriptions;
    /**
     * @type {?}
     * @protected
     */
    Package.prototype.packageId;
    /** @type {?} */
    Package.prototype.data;
    /**
     * @type {?}
     * @protected
     */
    Package.prototype.gsp;
}
