// @gsp-lcm/metadata-selector/esm5/metadata/metadata-select/components/selectmetadate/onclick.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/selectmetadate/onclick.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, HostListener, Output, EventEmitter, Renderer2, ElementRef } from '@angular/core';
import { throttleTime } from 'rxjs/operators';
import { Subject } from 'rxjs';
var OnceClickDirective = /** @class */ (function () {
    function OnceClickDirective(renderer, // Angular 2.x导入Renderer
    element) {
        this.renderer = renderer;
        this.element = element;
        // tslint:disable-next-line:no-output-rename
        this.clickCall = new EventEmitter();
        this.duration = 1000; // 必须是数字，传入时要用绑定语法
        // 必须是数字，传入时要用绑定语法
        this.$sub = new Subject();
    }
    /**
     * @return {?}
     */
    OnceClickDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // 如此绑定事件亦可
        // this.renderer.listen(
        //   this.element.nativeElement, 'click', event => {
        //     event.preventDefault();
        //     event.stopPropagation();
        //     this.$sub.next(event);
        //   }
        // );
        this.subscription = this.$sub.pipe(throttleTime(this.duration)).subscribe((/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            _this.clickCall.emit(e);
        }));
    };
    /**
     * @param {?} event
     * @return {?}
     */
    OnceClickDirective.prototype.clickEvent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault(); // 通常是不需要冒泡的
        event.stopPropagation();
        this.$sub.next(event);
    };
    /**
     * @return {?}
     */
    OnceClickDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subscription.unsubscribe();
    };
    OnceClickDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[click.once]'
                },] }
    ];
    /** @nocollapse */
    OnceClickDirective.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    OnceClickDirective.propDecorators = {
        clickCall: [{ type: Output, args: ['click.once',] }],
        duration: [{ type: Input }],
        clickEvent: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return OnceClickDirective;
}());
export { OnceClickDirective };
if (false) {
    /** @type {?} */
    OnceClickDirective.prototype.clickCall;
    /** @type {?} */
    OnceClickDirective.prototype.duration;
    /**
     * @type {?}
     * @private
     */
    OnceClickDirective.prototype.$sub;
    /**
     * @type {?}
     * @private
     */
    OnceClickDirective.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    OnceClickDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    OnceClickDirective.prototype.element;
}
