// @gsp-wf/wf-process-model/esm5/lib/model/designer/ParticipantSchemeForDesigner.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ParticipantSchemeForDesigner = /** @class */ (function () {
    function ParticipantSchemeForDesigner() {
        this.isAllScope = false;
        this.performers = [];
        this.variablePerformers = [];
        this.components = [];
        this.reportingLines = [];
        this.activityExecutors = [];
        this.matrixParticipants = [];
        this.organizationParticipants = new Array();
        this.options = {
            ExclusionPerformers: [],
            NoPerformerStrategy: '',
        };
    }
    return ParticipantSchemeForDesigner;
}());
export { ParticipantSchemeForDesigner };
if (false) {
    /** @type {?} */
    ParticipantSchemeForDesigner.prototype.performers;
    /** @type {?} */
    ParticipantSchemeForDesigner.prototype.variablePerformers;
    /** @type {?} */
    ParticipantSchemeForDesigner.prototype.components;
    /** @type {?} */
    ParticipantSchemeForDesigner.prototype.reportingLines;
    /** @type {?} */
    ParticipantSchemeForDesigner.prototype.activityExecutors;
    /** @type {?} */
    ParticipantSchemeForDesigner.prototype.matrixParticipants;
    /** @type {?} */
    ParticipantSchemeForDesigner.prototype.organizationParticipants;
    /** @type {?} */
    ParticipantSchemeForDesigner.prototype.options;
    /** @type {?} */
    ParticipantSchemeForDesigner.prototype.isAllScope;
}
var ParticipantItem = /** @class */ (function () {
    function ParticipantItem() {
    }
    return ParticipantItem;
}());
export { ParticipantItem };
if (false) {
    /** @type {?} */
    ParticipantItem.prototype.id;
    /** @type {?} */
    ParticipantItem.prototype.performerId;
    /** @type {?} */
    ParticipantItem.prototype.performerName;
    /** @type {?} */
    ParticipantItem.prototype.positionType;
    /** @type {?} */
    ParticipantItem.prototype.orgId;
    /** @type {?} */
    ParticipantItem.prototype.orgName;
    /** @type {?} */
    ParticipantItem.prototype.orgType;
    /** @type {?} */
    ParticipantItem.prototype.type;
    /** @type {?} */
    ParticipantItem.prototype.condition;
    /** @type {?} */
    ParticipantItem.prototype.tag;
    /** @type {?} */
    ParticipantItem.prototype.varValue;
    /** @type {?} */
    ParticipantItem.prototype.varType;
    /** @type {?} */
    ParticipantItem.prototype.relation;
}
var OrganizationParticipant = /** @class */ (function () {
    function OrganizationParticipant() {
    }
    return OrganizationParticipant;
}());
export { OrganizationParticipant };
if (false) {
    /** @type {?} */
    OrganizationParticipant.prototype.id;
    /** @type {?} */
    OrganizationParticipant.prototype.name;
    /** @type {?} */
    OrganizationParticipant.prototype.actionScope;
    /** @type {?} */
    OrganizationParticipant.prototype.organizationLevel;
    /** @type {?} */
    OrganizationParticipant.prototype.organizationId;
    /** @type {?} */
    OrganizationParticipant.prototype.organizationCode;
    /** @type {?} */
    OrganizationParticipant.prototype.organizationName;
    /** @type {?} */
    OrganizationParticipant.prototype.includeChildren;
    /** @type {?} */
    OrganizationParticipant.prototype.parentsUntil;
    /** @type {?} */
    OrganizationParticipant.prototype.tag;
    /** @type {?} */
    OrganizationParticipant.prototype.condition;
}
