// @gsp-svc/file-viewer/esm2015/lib/excel-document/excel-document.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, ViewChildren, QueryList, Renderer2 } from '@angular/core';
import { ExcelPageComponent } from '../excel-page/excel-page.component';
import { NavigateService } from '../service/navigate.service';
import { ZoomService } from '../service/zoom.service';
import { WindowService } from '../service/window.service';
import { GspDocumentComponent } from '../gsp-document/gsp-document/gsp-document.component';
import { DocViewerService } from '../service/doc-viewer.service';
export class ExcelDocumentComponent extends GspDocumentComponent {
    //@Input() file: FileDescription;
    /**
     * @param {?} _elementRef
     * @param {?} zoomService
     * @param {?} windowService
     * @param {?} navigateService
     * @param {?} viewerService
     * @param {?} renderer
     */
    constructor(_elementRef, zoomService, windowService, navigateService, viewerService, renderer) {
        super(_elementRef, zoomService, windowService, navigateService);
        this.viewerService = viewerService;
        this.renderer = renderer;
        this.panzoom = null;
        this.navigateService = navigateService;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.currentPageNo = 1;
        document.getElementById("excelContent").style.height = document.getElementById("paneContent").offsetHeight - 23 + 'px';
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.refreshExcelDocHeight();
        this.pages.changes.subscribe((/**
         * @return {?}
         */
        () => {
            this.refreshExcelDocHeight();
        }));
        this.navigateService.navigate.subscribe((((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            if (value) {
                this.selectSheet(value);
            }
        }))));
        /** @type {?} */
        const scrollbarWidth = this.getScrollBarWidth();
        this.renderer.setStyle(this._elementRef.nativeElement.querySelector('.sheets'), 'right', this.getScrollBarWidth() + 'px');
        this.renderer.setStyle(this._elementRef.nativeElement.querySelector('.sheets'), 'bottom', this.getScrollBarWidth() + 'px');
        if (scrollbarWidth === 0) {
            this.renderer.setStyle(this._elementRef.nativeElement.querySelector('.sheets'), 'padding-right', '17px');
        }
    }
    /**
     * @return {?}
     */
    getScrollBarWidth() {
        /** @type {?} */
        const documentBox = (/** @type {?} */ (document.querySelector('.app-gsp-document')));
        /** @type {?} */
        const scrollbarWidth = documentBox.offsetWidth - documentBox.clientWidth;
        return scrollbarWidth;
    }
    /**
     * @return {?}
     */
    refreshExcelDocHeight() {
        // For current iteration we'll change actual height of .document
        this.doc = this._elementRef.nativeElement.children.item(0);
        this.panzoom = this._elementRef.nativeElement.children.item(0).children.item(0);
        // magic number 37 is the height of the bottom-bar with navigation between pages
        this.doc.style.height = this.panzoom.scrollHeight + 37 + "px";
    }
    /**
     * @param {?} number
     * @return {?}
     */
    selectSheet(number) {
        this.currentPageNo = number;
        if (this.file.pages[number - 1].data == "") {
            /** @type {?} */
            let credentials = { guid: this.file.filePath, password: "" };
            this.viewerService.loadPage(credentials, number).subscribe((/**
             * @param {?} page
             * @return {?}
             */
            (page) => {
                this.file.pages[number - 1] = page;
            }));
        }
    }
}
ExcelDocumentComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-excel-document',
                template: "<!-- <div class=\"wait\" *ngIf=\"wait\">Please wait...</div> -->\r\n<div class=\"f-utils-fill-flex-column\" style=\"height: 100%;\">\r\n    <div style=\"height: 100%;\">\r\n        <div style=\"background-color: white; overflow: auto;\">\r\n            <div id=\"document f-utils-fill\" class=\"document\">\r\n                <div id=\"excelContent\" [ngClass]=\"isDesktop ? 'panzoom auto-height' : 'panzoom mobile'\" style=\"overflow: auto;\" gdZoom [zoomActive]=\"true\" [file]=\"file\" gdSearchable>\r\n                    <div [ngClass]=\"file.showGridLines ? 'page-grid-lines' : 'page'\" *ngFor=\"let page of file?.pages\" gdRotation [angle]=\"page.angle\" [isHtmlMode]=\"mode\">\r\n                        <app-excel-page *ngIf=\"currentPageNo == page.number\" [number]=\"page.number\" [data]=\"page.data\" [isHtml]=\"mode\" [angle]=\"page.angle\" [width]=\"page.width\" [height]=\"page.height\" [editable]=\"page.editable\">\r\n                        </app-excel-page>\r\n                    </div>\r\n                </div>\r\n            </div>\r\n        </div>\r\n        <div class=\"sheets\" style=\"position: absolute; bottom: 0px;\">\r\n            <div class=\"sheets-wrapper\">\r\n                <div *ngFor=\"let page of file?.pages\">\r\n                    <button style=\"background-color: white; border: 0;\" [ngClass]=\"{'active': currentPageNo == page.number }\" (click)=\"selectSheet(page.number)\">{{page.name}}</button>\r\n                </div>\r\n            </div>\r\n        </div>\r\n    </div>\r\n</div>",
                styles: [":host{overflow:scroll;width:100%;background-color:#e7e7e7}.document{width:100%;transition:.4s;padding:0;margin:0;position:relative}.sheets{background-color:#fff;display:flex;border-top:1px solid #e7e7e7;position:fixed;width:100%}.sheets ::ng-deep gd-button.active .text{background-color:#272727;border-radius:10px;color:#eee}.sheets ::ng-deep gd-button .text{padding:1px 12px;color:#000}.sheets ::ng-deep gd-button fa-icon{display:none}.sheets-wrapper{margin-left:29px;display:flex}.page{position:relative;display:inline-block;background-color:#fff;transition:.3s}.wait{position:absolute;top:55px;left:Calc(30%)}.panzoom.auto-height{-webkit-transform:none;transform:none;-webkit-backface-visibility:hidden;backface-visibility:hidden;-webkit-transform-origin:50% 50% 0;transform-origin:50% 50% 0;display:flex;flex-wrap:wrap}.gd-zoomed{margin:10px 98px}.highlights{position:absolute;top:0;left:0;bottom:0;right:0}.page-grid-lines{background-color:#fff}@media (max-width:1037px){.document{overflow-x:auto!important}.page{min-width:unset!important;min-height:unset!important;margin:5px 0}}"]
            }] }
];
/** @nocollapse */
ExcelDocumentComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: ZoomService },
    { type: WindowService },
    { type: NavigateService },
    { type: DocViewerService },
    { type: Renderer2 }
];
ExcelDocumentComponent.propDecorators = {
    pages: [{ type: ViewChildren, args: [ExcelPageComponent,] }]
};
if (false) {
    /** @type {?} */
    ExcelDocumentComponent.prototype.pages;
    /** @type {?} */
    ExcelDocumentComponent.prototype.currentPageNo;
    /** @type {?} */
    ExcelDocumentComponent.prototype.panzoom;
    /** @type {?} */
    ExcelDocumentComponent.prototype.navigateService;
    /** @type {?} */
    ExcelDocumentComponent.prototype.doc;
    /**
     * @type {?}
     * @private
     */
    ExcelDocumentComponent.prototype.viewerService;
    /**
     * @type {?}
     * @private
     */
    ExcelDocumentComponent.prototype.renderer;
}
