// @farris/dynamic-control-group/esm2015/lib/components/number/number-range.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from "@angular/core";
import { FormGroup } from "@angular/forms";
export class NumberRangeComponent {
    constructor() { }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.options = this.control.control;
        this.className = this.control.className ? this.control.className : 'col-12 col-md-6 col-xl-3 col-el-2';
    }
}
NumberRangeComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-number-range",
                template: `
    <div class="farris-group-wrap" [formGroup]="form">
      <div class="form-group farris-form-group">
        <label class="col-form-label">
          <span *ngIf="0" class="farris-label-info text-danger">*</span>
          <span class="farris-label-text">{{control.name}}</span>
        </label>
        <div class="farris-input-wrap">
            <farris-number-range
                [formControlName]="control.labelCode"
                [max]="options.max"
                [min]="options.min"
                [precision]="options.precision"
                [readonly]="options.readonly"
            >
            </farris-number-range>
        </div>
      </div>
    </div>
  `
            }] }
];
/** @nocollapse */
NumberRangeComponent.ctorParameters = () => [];
NumberRangeComponent.propDecorators = {
    className: [{ type: HostBinding, args: ['class',] }],
    control: [{ type: Input }],
    form: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NumberRangeComponent.prototype.className;
    /** @type {?} */
    NumberRangeComponent.prototype.control;
    /** @type {?} */
    NumberRangeComponent.prototype.form;
    /** @type {?} */
    NumberRangeComponent.prototype.options;
}
