// @farris/ui-shortcuts/fesm2015/farris-ui-shortcuts.js
import { Subject } from 'rxjs';
import { filter } from 'rxjs/operators';
import { Injectable, NgModule, defineInjectable } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const isMac = (navigator.userAgent.indexOf('Mac OS') !== -1);
/** @type {?} */
const codes = {
    'backspace': 8,
    'tab': 9,
    'enter': 13,
    'pause/break': 19,
    'caps lock': 20,
    'esc': 27,
    'space': 32,
    'page up': 33,
    'page down': 34,
    'end': 35,
    'home': 36,
    'left': 37,
    'up': 38,
    'right': 39,
    'down': 40,
    'insert': 45,
    'delete': 46,
    'numpad *': 106,
    'numpad +': 107,
    'numpad -': 109,
    'numpad .': 110,
    'numpad /': 111,
    'num lock': 144,
    'scroll lock': 145,
    'my computer': 182,
    'my calculator': 183,
    ';': 186,
    '=': 187,
    ',': 188,
    '-': 189,
    '.': 190,
    '/': 191,
    '`': 192,
    '[': 219,
    '\\': 220,
    ']': 221,
    "'": 222,
    'windows': 91,
    '⇧': 16,
    '⌥': 18,
    '⌃': 17,
    'pause': 19,
    'break': 19,
    'caps': 20,
    'return': 13,
    'escape': 27,
    'spc': 32,
    'pgup': 33,
    'pgdn': 34,
    'ins': 45,
    'del': 46,
    'F2': 113
};
/** @type {?} */
const modifiers = {
    'shift': 'shiftKey',
    'ctrl': 'ctrlKey',
    'alt': 'altKey',
    'cmd': isMac ? 'metaKey' : 'ctrlKey',
    'command': isMac ? 'metaKey' : 'ctrlKey',
    'left command': 'metaKey',
    'right command': 'MetaRight',
    '⌘': isMac ? 'metaKey' : 'ctrlKey',
    'option': 'altKey',
    'ctl': 'ctrlKey',
    'control': 'ctrlKey',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const any = (/**
 * @param {?} fn
 * @param {?} list
 * @return {?}
 */
(fn, list) => {
    /** @type {?} */
    let idx = 0;
    while (idx < list.length) {
        if (fn(list[idx])) {
            return true;
        }
        idx += 1;
    }
    return false;
});
/** @type {?} */
const identity = (/**
 * @param {?} x
 * @return {?}
 */
x => x);
/** @type {?} */
const isNill = (/**
 * @param {?} x
 * @return {?}
 */
x => x == null);
/** @type {?} */
const difference = (/**
 * @param {?} first
 * @param {?} second
 * @return {?}
 */
(first, second) => first.filter((/**
 * @param {?} item
 * @return {?}
 */
item => !((/** @type {?} */ (second))).some((/**
 * @param {?} secondItem
 * @return {?}
 */
secondItem => secondItem == item)))));
/** @type {?} */
const allPass = (/**
 * @param {?} preds
 * @return {?}
 */
preds => (/**
 * @param {...?} args
 * @return {?}
 */
(...args) => {
    /** @type {?} */
    let idx = 0;
    /** @type {?} */
    const len = preds.length;
    while (idx < len) {
        if (!preds[idx].apply(this, args)) {
            return false;
        }
        idx += 1;
    }
    return true;
}));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const $$ngOnDestroy = Symbol('OnDestroy');
class ShortcutsService {
    constructor() {
        /**
         * 解析快捷键
         * 并为每个KEY 创建一个函数
         */
        this._shortcuts = [];
        /**
         * 控制按键时间.
         */
        this.throttleTime = 0;
        this._pressed = new Subject();
        /**
         * Streams of pressed events, can be used instead or with a command.
         */
        this.pressed$ = this._pressed.asObservable();
        /**
         * 禁用快捷键
         */
        this.disabled = false;
        this._ignored = ['INPUT', 'TEXTAREA', 'SELECT'];
        this.isAllowed = (/**
         * @param {?} shortcut
         * @return {?}
         */
        (shortcut) => {
            /** @type {?} */
            const target = (/** @type {?} */ (shortcut.event.target));
            if (target === shortcut.target) {
                return true;
            }
            if (shortcut.allowIn.length) {
                return !((/** @type {?} */ (difference(this._ignored, shortcut.allowIn)))).includes(target.nodeName);
            }
            return !((/** @type {?} */ (this._ignored))).includes(target.nodeName);
        });
        this.mapEvent = (/**
         * @param {?} event
         * @return {?}
         */
        event => this._shortcuts
            .map((/**
         * @param {?} shortcut
         * @return {?}
         */
        shortcut => Object.assign({}, shortcut, {
            predicates: any(identity, shortcut.predicates.map((/**
             * @param {?} predicates
             * @return {?}
             */
            (predicates) => allPass(predicates)(event)))),
            event: event
        })))
            .filter((/**
         * @param {?} shortcut
         * @return {?}
         */
        shortcut => shortcut.predicates))
            .reduce((/**
         * @param {?} acc
         * @param {?} shortcut
         * @return {?}
         */
        (acc, shortcut) => (acc.priority > shortcut.priority ? acc : shortcut)), (/** @type {?} */ ({
            priority: 0
        }))));
        /**
         * transforms a shortcut to:
         * a predicate function
         */
        this.getKeys = (/**
         * @param {?} command
         * @return {?}
         */
        (command) => command.map((/**
         * @param {?} key
         * @return {?}
         */
        key => key.trim())).filter((/**
         * @param {?} key
         * @return {?}
         */
        key => key !== '+'))
            .map((/**
         * @param {?} key
         * @return {?}
         */
        key => {
            // for modifiers like control key
            // look for event['ctrlKey']
            // otherwise use the keyCode
            if (modifiers.hasOwnProperty(key)) {
                return (/**
                 * @param {?} event
                 * @return {?}
                 */
                event => !!event[modifiers[key]]);
            }
            return (/**
             * @param {?} event
             * @return {?}
             */
            event => codes[key]
                ? event.keyCode === codes[key] || event.key === key
                : event.keyCode === key.toUpperCase().charCodeAt(0));
        })));
        // this.keydown$ = fromEvent(document, 'keydown').pipe(
        //     filter(_ => !this.disabled),
        //     map(this.mapEvent),
        //     filter(
        //         (shortcut: ParsedShortcut) =>
        //             !shortcut.target || shortcut.event.target === shortcut.target
        //     ),
        //     filter((shortcut: ParsedShortcut) => isFunction(shortcut.command)),
        //     filter(this.isAllowed),
        //     tap((shortcut: any) => !shortcut.preventDefault || shortcut.event.preventDefault()),
        //     debounce(shortcut => timer(shortcut.throttleTime)),
        //     tap(shortcut => shortcut.command({ event: shortcut.event, key: shortcut.key })),
        //     tap(shortcut => this._pressed.next({ event: shortcut.event, key: shortcut.key })),
        //     catchError(error => of(error))
        // );
        // this.subscription = this.keydown$.subscribe();
    }
    /**
     * @private
     * @return {?}
     */
    get shortcuts() {
        return this._shortcuts;
    }
    /**
     * Remove subscription.
     * @return {?}
     */
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
    /**
     * Enable all keyboard shortcuts
     * @return {?}
     */
    enable() {
        this.disabled = false;
        return this;
    }
    /**
     * Disable all keyboard shortcuts
     * @return {?}
     */
    disable() {
        this.disabled = true;
        return this;
    }
    /**
     * Check if all keyboard shortcuts are disabled.
     * @return {?}
     */
    isDisabled() {
        return this.disabled;
    }
    /**
     * Add new shortcut/s
     * @param {?} shortcuts
     * @param {?=} instance
     * @return {?}
     */
    add(shortcuts, instance) {
        shortcuts = Array.isArray(shortcuts) ? shortcuts : [shortcuts];
        if (instance) {
            const [key] = [...shortcuts.map((/**
                 * @param {?} shortcut
                 * @return {?}
                 */
                shortcut => shortcut.key))];
            this.bindOnDestroy(instance, key);
        }
        this._shortcuts.push(...this.parseCommand(shortcuts));
        return this;
    }
    /**
     * bind to the component ngOnDestroy to remove related keys
     * when component is destroyed.
     * @private
     * @param {?} instance - component to remove keys when ngOnDestroy is called.
     * @param {?} keys
     * @return {?}
     */
    bindOnDestroy(instance, keys) {
        if (instance.ngOnDestroy) {
            instance[$$ngOnDestroy] = instance.ngOnDestroy;
        }
        /** @type {?} */
        const that = this;
        instance.ngOnDestroy = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            const onDestroy = instance[$$ngOnDestroy];
            if (onDestroy) {
                onDestroy.apply(this);
            }
            that.remove(keys);
        });
        return this;
    }
    /**
     * Remove a command based on key or array of keys.
     * can be used for cleanup.
     * @param {?} key
     * @return {?}
     */
    remove(key) {
        /** @type {?} */
        const keys = Array.isArray(key) ? key : [key];
        this._shortcuts = this._shortcuts.filter((/**
         * @param {?} shortcut
         * @return {?}
         */
        shortcut => {
            return !shortcut.key.find((/**
             * @param {?} sKey
             * @return {?}
             */
            sKey => {
                return keys.filter((/**
                 * @param {?} k
                 * @return {?}
                 */
                k => k === sKey)).length > 0;
            }));
        }));
        return this;
    }
    /**
     * Returns an observable of keyboard shortcut filtered by a specific key.
     * @param {?} key - the key to filter the observable by.
     * @return {?}
     */
    select(key) {
        return this.pressed$.pipe(filter((/**
         * @param {?} __0
         * @return {?}
         */
        ({ event, key: eventKeys }) => {
            return !!eventKeys.find((/**
             * @param {?} eventKey
             * @return {?}
             */
            eventKey => eventKey === key));
        })));
    }
    /**
     * Parse each command using getKeys function
     * @private
     * @param {?} command
     * @return {?}
     */
    parseCommand(command) {
        /** @type {?} */
        const commands = Array.isArray(command) ? command : [command];
        return commands.map((/**
         * @param {?} cmd
         * @return {?}
         */
        (cmd) => {
            /** @type {?} */
            const keys = Array.isArray(cmd.key) ? cmd.key : [cmd.key];
            /** @type {?} */
            const priority = Math.max(...keys.map((/**
             * @param {?} key
             * @return {?}
             */
            key => key.split(' ').length)));
            /** @type {?} */
            const predicates = keys.map((/**
             * @param {?} key
             * @return {?}
             */
            key => this.getKeys(key.split(' '))));
            return (/** @type {?} */ (Object.assign({}, cmd, { allowIn: cmd.allowIn || [], key: keys, throttle: isNill(cmd.throttleTime) ? this.throttleTime : cmd.throttleTime, priority: priority, predicates: predicates })));
        }));
    }
}
ShortcutsService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ShortcutsService.ctorParameters = () => [];
/** @nocollapse */ ShortcutsService.ngInjectableDef = defineInjectable({ factory: function ShortcutsService_Factory() { return new ShortcutsService(); }, token: ShortcutsService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ShortcutsModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: ShortcutsModule,
            providers: [
                ShortcutsService
            ]
        };
    }
}
ShortcutsModule.decorators = [
    { type: NgModule, args: [{
                declarations: [],
                imports: [],
                exports: []
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { ShortcutsService, ShortcutsModule };
