// @edp-aif/runtime-api/esm5/lib/entity/BefContentVariable.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { CreationRuleExecutorAssociations } from './CreationRuleExecutorAssociations';
var BefContentVariable = /** @class */ (function (_super) {
    tslib_1.__extends(BefContentVariable, _super);
    function BefContentVariable() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.associations = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    BefContentVariable.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.upperBeFuncSessionId = null;
        this.associations = new Array();
    };
    /**
     * @return {?}
     */
    BefContentVariable.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upperBeFuncSessionId", this.upperBeFuncSessionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BefContentVariable.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.upperBeFuncSessionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upperBeFuncSessionId");
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", CreationRuleExecutorAssociations);
    };
    return BefContentVariable;
}(JsonConvertableImpl));
export { BefContentVariable };
if (false) {
    /** @type {?} */
    BefContentVariable.prototype.upperBeFuncSessionId;
    /** @type {?} */
    BefContentVariable.prototype.associations;
}
