// @edp-aif/runtime-api/esm5/lib/entity/CreationRuleExecutorArgs.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from '@edp-aif/common-api';
var CreationRuleExecutorArgs = /** @class */ (function (_super) {
    tslib_1.__extends(CreationRuleExecutorArgs, _super);
    function CreationRuleExecutorArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.sourceVoEntities = new Array();
        _this.targetVoProxyEntities = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    CreationRuleExecutorArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.sourceVoEntities = new Array();
        this.creationRuleId = null;
        this.bizFlowchartId = null;
        this.sourceVoId = null;
        this.targetVoProxyId = null;
        this.targetVoProxyEntities = new Array();
    };
    /**
     * @return {?}
     */
    CreationRuleExecutorArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntities", this.sourceVoEntities);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyId", this.targetVoProxyId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyEntities", this.targetVoProxyEntities);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CreationRuleExecutorArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.sourceVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.sourceVoEntities = BpmnModelJsonHelper.GetAttributeValue(obj, "sourceVoEntities");
        this.targetVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyId");
        this.targetVoProxyEntities = BpmnModelJsonHelper.GetAttributeValue(obj, "targetVoProxyEntities");
    };
    return CreationRuleExecutorArgs;
}(JsonConvertableImpl));
export { CreationRuleExecutorArgs };
if (false) {
    /** @type {?} */
    CreationRuleExecutorArgs.prototype.creationRuleId;
    /** @type {?} */
    CreationRuleExecutorArgs.prototype.sourceVoId;
    /** @type {?} */
    CreationRuleExecutorArgs.prototype.bizFlowchartId;
    /** @type {?} */
    CreationRuleExecutorArgs.prototype.sourceVoEntities;
    /** @type {?} */
    CreationRuleExecutorArgs.prototype.targetVoProxyId;
    /** @type {?} */
    CreationRuleExecutorArgs.prototype.targetVoProxyEntities;
}
