// @edp-aif/runtime-api/esm5/lib/entity/CreationRuleExecutorResult.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { CreationMappingResult } from './CreationMappingResult';
var CreationRuleExecutorResult = /** @class */ (function (_super) {
    tslib_1.__extends(CreationRuleExecutorResult, _super);
    function CreationRuleExecutorResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isAllAssociation = false;
        _this.mappingResults = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    CreationRuleExecutorResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.creationRuleId = null;
        this.creationRuleInstanceId = null;
        this.bizFlowchartId = null;
        this.isAllAssociation = false;
        this.mappingResults = new Array();
    };
    /**
     * @return {?}
     */
    CreationRuleExecutorResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isAllAssociation", this.isAllAssociation);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingResults", this.mappingResults);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CreationRuleExecutorResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.creationRuleInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.isAllAssociation = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isAllAssociation");
        this.mappingResults = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingResults", CreationMappingResult);
    };
    return CreationRuleExecutorResult;
}(JsonConvertableImpl));
export { CreationRuleExecutorResult };
if (false) {
    /** @type {?} */
    CreationRuleExecutorResult.prototype.creationRuleId;
    /** @type {?} */
    CreationRuleExecutorResult.prototype.creationRuleInstanceId;
    /** @type {?} */
    CreationRuleExecutorResult.prototype.bizFlowchartId;
    /** @type {?} */
    CreationRuleExecutorResult.prototype.isAllAssociation;
    /** @type {?} */
    CreationRuleExecutorResult.prototype.mappingResults;
}
