// @edp-aif/runtime-api/esm5/lib/entity/CreationMappingResult.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { MappingResult } from './MappingResult';
import { MappingAssociation } from './MappingAssociation';
import { BpmnModelJsonHelper } from '@edp-aif/common-api';
var CreationMappingResult = /** @class */ (function (_super) {
    tslib_1.__extends(CreationMappingResult, _super);
    function CreationMappingResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.associations = new Array();
        _this.entityData = null;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    CreationMappingResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.associations = new Array();
        this.entityData = null;
    };
    /**
     * @return {?}
     */
    CreationMappingResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        BpmnModelJsonHelper.SetAttributeValue(obj, "entityData", this.entityData);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CreationMappingResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
        this.entityData = BpmnModelJsonHelper.GetAttributeValue(obj, "entityData");
    };
    return CreationMappingResult;
}(MappingResult));
export { CreationMappingResult };
if (false) {
    /** @type {?} */
    CreationMappingResult.prototype.associations;
    /** @type {?} */
    CreationMappingResult.prototype.entityData;
}
