// @edp-aif/runtime-api/esm5/lib/entity/CreationRuleExecutorAssociations.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { MappingAssociation } from './MappingAssociation';
var CreationRuleExecutorAssociations = /** @class */ (function (_super) {
    tslib_1.__extends(CreationRuleExecutorAssociations, _super);
    function CreationRuleExecutorAssociations() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isAllAssociation = false;
        _this.associations = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    CreationRuleExecutorAssociations.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.creationRuleId = null;
        this.creationRuleInstanceId = null;
        this.bizFlowchartId = null;
        this.creationProcessId = null;
        this.creationProcessInstanceId = null;
        this.isAllAssociation = false;
        this.associations = new Array();
    };
    /**
     * @return {?}
     */
    CreationRuleExecutorAssociations.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessId", this.creationProcessId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessInstanceId", this.creationProcessInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isAllAssociation", this.isAllAssociation);
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CreationRuleExecutorAssociations.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.creationRuleInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.creationProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessId");
        this.creationProcessInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessInstanceId");
        this.isAllAssociation = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isAllAssociation");
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
    };
    return CreationRuleExecutorAssociations;
}(JsonConvertableImpl));
export { CreationRuleExecutorAssociations };
if (false) {
    /** @type {?} */
    CreationRuleExecutorAssociations.prototype.creationRuleId;
    /** @type {?} */
    CreationRuleExecutorAssociations.prototype.creationRuleInstanceId;
    /** @type {?} */
    CreationRuleExecutorAssociations.prototype.bizFlowchartId;
    /** @type {?} */
    CreationRuleExecutorAssociations.prototype.creationProcessId;
    /** @type {?} */
    CreationRuleExecutorAssociations.prototype.creationProcessInstanceId;
    /** @type {?} */
    CreationRuleExecutorAssociations.prototype.isAllAssociation;
    /** @type {?} */
    CreationRuleExecutorAssociations.prototype.associations;
}
