// @edp-aif/runtime-api/esm5/lib/entity/FormBillCreationProcessExecutorResult.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper, PushCreationMode } from "@edp-aif/common-api";
import { ConditionResult } from './ConditionResult';
import { RpcCreationRuleExecutorResult } from "./rpc/RpcCreationRuleExecutorResult";
var FormBillCreationProcessExecutorResult = /** @class */ (function (_super) {
    tslib_1.__extends(FormBillCreationProcessExecutorResult, _super);
    function FormBillCreationProcessExecutorResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.pushCreationMode = PushCreationMode.CREATION;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormBillCreationProcessExecutorResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.conditionResult = null;
        this.creationRuleExecutorResult = null;
        this.pushCreationMode = PushCreationMode.CREATION;
        this.downBillSuCode = null;
    };
    /**
     * @return {?}
     */
    FormBillCreationProcessExecutorResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "conditionResult", this.conditionResult);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleExecutorResult", this.creationRuleExecutorResult);
        BpmnModelJsonHelper.SetAttributeValue(obj, "pushCreationMode", this.pushCreationMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillSuCode", this.downBillSuCode);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormBillCreationProcessExecutorResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.conditionResult = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "conditionResult", ConditionResult);
        this.creationRuleExecutorResult = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "creationRuleExecutorResult", RpcCreationRuleExecutorResult);
        /** @type {?} */
        var pushCreationMode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "pushCreationMode");
        if (pushCreationMode != null) {
            this.pushCreationMode = PushCreationMode[pushCreationMode];
        }
        this.downBillSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillSuCode");
    };
    return FormBillCreationProcessExecutorResult;
}(JsonConvertableImpl));
export { FormBillCreationProcessExecutorResult };
if (false) {
    /** @type {?} */
    FormBillCreationProcessExecutorResult.prototype.conditionResult;
    /** @type {?} */
    FormBillCreationProcessExecutorResult.prototype.creationRuleExecutorResult;
    /** @type {?} */
    FormBillCreationProcessExecutorResult.prototype.pushCreationMode;
    /** @type {?} */
    FormBillCreationProcessExecutorResult.prototype.downBillSuCode;
}
