// @edp-aif/runtime-api/esm5/lib/entity/MappingResult.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
var MappingResult = /** @class */ (function (_super) {
    tslib_1.__extends(MappingResult, _super);
    function MappingResult() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    MappingResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.voId = null;
        this.changeDetail = null;
    };
    /**
     * @return {?}
     */
    MappingResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "changeDetail", this.changeDetail);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    MappingResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
        this.changeDetail = JSON.parse(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "changeDetail"));
    };
    return MappingResult;
}(JsonConvertableImpl));
export { MappingResult };
if (false) {
    /** @type {?} */
    MappingResult.prototype.voId;
    /** @type {?} */
    MappingResult.prototype.changeDetail;
}
