// @gsp-bef/gsp-vo-metadata/esm2015/lib/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// export * from './view-model';
// export * from './view-model/ref';
export { GSPViewModel } from './view-model/src/GSPViewModel';
export { GSPViewModelElement } from './view-model/src/GSPViewModelElement';
export { GSPViewObject } from './view-model/src/GSPViewObject';
export { VoDataExtendInfo, VoDataExtendInfos } from './view-model/src/VoDataExtendInfo/VoDataExtendInfo';
export { MappedBizAction } from './view-model/src/Action/MappedBizAction';
export { MappedCdpAction } from './view-model/src/Action/MappedCdpAction';
export { ViewModelAction } from './view-model/src/Action/ViewModelAction';
export { VMActionResult } from './view-model/src/Action/VMActionResult';
export { TemplateVoInfo } from './view-model/src/Common/TemplateVoInfo';
export { ViewModelParameter } from './view-model/src/Action/Para/ViewModelParameter';
export { MappedBizActionParameter } from './view-model/src/Action/Para/MappedBizActionParameter';
export { MappedCdpActionParameter } from './view-model/src/Action/Para/MappedCdpActionParameter';
export { VMParameterMode, VMParameterModes } from './view-model/src/Action/ParaEnum/VMParameterMode';
export { VMParameterType, VMParameterTypes } from './view-model/src/Action/ParaEnum/VMParameterType';
export { VMCollectionParameterType, VMCollectionParameterTypes } from './view-model/src/Action/ParaEnum/VMCollectionParameterType';
export { ViewModelReturnValue } from './view-model/src/Action/ReturnValue/ViewModelReturnValue';
export { ViewModelVoidReturnValue } from './view-model/src/Action/ReturnValue/ViewModelVoidReturnValue';
export { ValueHelpConfig } from './view-model/src/Common/ValueHelpConfig';
export { ViewModelMapping } from './view-model/src/Common/ViewModelMapping';
export { HelpExtendAction } from './view-model/src/Common/HelpExtendAction';
export { GspVoElementMapping } from './view-model/src/Common/GspVoElementMapping';
export { GspVoObjectMapping } from './view-model/src/Common/GspVoObjectMapping';
export { ElementHelpType } from './view-model/src/Enum/ElementHelpType';
export { MappingType, MappingTypes } from './view-model/src/Enum/MappingType';
export { MetadataType, MetadataTypes } from './view-model/src/Enum/MetadataType';
export { ViewModelActionType, ViewModelActionTypes } from './view-model/src/Enum/ViewModelActionType';
export { GspVoObjectSourceType } from './view-model/src/Enum/GspVoObjectSourceType';
export { GspVoElementSourceType } from './view-model/src/Enum/GspVoElementSourceType';
