// @farris/mobile-vue-adapter/bundles/farris-mobile-vue-adapter.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@farris/mobile-devkit'), require('vue'), require('@farris/mobile-command-services'), require('@farris/mobile-bef'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@farris/mobile-vue-adapter', ['exports', '@farris/mobile-devkit', 'vue', '@farris/mobile-command-services', '@farris/mobile-bef', 'rxjs', 'rxjs/operators'], factory) :
    (global = global || self, factory((global.farris = global.farris || {}, global.farris['mobile-vue-adapter'] = {}), global.mobileDevkit, global.vue, global.mobileCommandServices, global.mobileBef, global.rxjs, global.rxjs.operators));
}(this, (function (exports, mobileDevkit, vue, mobileCommandServices, mobileBef, rxjs, operators) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/recompute.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 包装计算数据getter方法
     * @param {?} fn
     * @param {?} propName
     * @return {?}
     */
    function recomputable(fn, propName) {
        /** @type {?} */
        var reactiveObj = vue.reactive({
            stateVersion: 0
        });
        return (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var self = this;
            if (!self.$__recomputables) {
                self.$__recomputables = {};
            }
            if (!self.$__recomputables[propName]) {
                self.$__recomputables[propName] = reactiveObj;
            }
            /** @type {?} */
            var stateVersion = reactiveObj.stateVersion;
            return fn.call(self);
        });
    }
    /**
     * 触发计算属性重新计算
     * @param {?} vm
     * @param {?} propName
     * @return {?}
     */
    function recompute(vm, propName) {
        if (!vm.$__recomputables || !vm.$__recomputables[propName]) {
            return;
        }
        vm.$__recomputables[propName].stateVersion++;
        try {
            vm.$forceUpdate();
        }
        catch (error) {
            console.log('$forceUpdate', error);
        }
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/binding-states.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 绑定状态
     */
    var   /**
     * 绑定状态
     */
    BindingStates = /** @class */ (function () {
        /**
         * 构造函数
         * @param viewModelContext ViewModel上下文
         */
        function BindingStates(vueInstance) {
            this.vueInstance = vueInstance;
            /** @type {?} */
            var viewModel = (/** @type {?} */ (this.vueInstance.$getViewModel()));
            this.viewModelContext = viewModel.context;
            this.subscriptionsMap = new Map();
        }
        /**
         * 映射实体列表数据
         */
        /**
         * 映射实体列表数据
         * @return {?}
         */
        BindingStates.prototype.getEntityListData = /**
         * 映射实体列表数据
         * @return {?}
         */
        function () {
            var _this = this;
            /** @type {?} */
            var stateName = 'entityListData';
            if (this.subscriptionsMap.has(stateName) === false) {
                /** @type {?} */
                var subscription = this.viewModelContext.bindingData.changes.subscribe((/**
                 * @param {?} change
                 * @return {?}
                 */
                function (change) {
                    recompute(_this.vueInstance, stateName);
                }));
                this.subscriptionsMap.set(stateName, subscription);
            }
            return this.viewModelContext.bindingData.getList().toJSON();
        };
        /**
         * 映射当前实体数据
         */
        /**
         * 映射当前实体数据
         * @return {?}
         */
        BindingStates.prototype.getCurrentEntityData = /**
         * 映射当前实体数据
         * @return {?}
         */
        function () {
            var _this = this;
            /** @type {?} */
            var stateName = 'currentEntityData';
            if (this.subscriptionsMap.has(stateName) === false) {
                /** @type {?} */
                var subscription = this.viewModelContext.bindingData.changes.subscribe((/**
                 * @param {?} change
                 * @return {?}
                 */
                function (change) {
                    recompute(_this.vueInstance, stateName);
                }));
                this.subscriptionsMap.set(stateName, subscription);
            }
            return this.viewModelContext.bindingData.getObject().toJSON();
        };
        /**
         * 映射UIState数据
         */
        /**
         * 映射UIState数据
         * @return {?}
         */
        BindingStates.prototype.getUIState = /**
         * 映射UIState数据
         * @return {?}
         */
        function () {
            var _this = this;
            /** @type {?} */
            var stateName = 'uiState';
            if (this.subscriptionsMap.has(stateName) === false) {
                /** @type {?} */
                var subscription = this.viewModelContext.uiState.changes.subscribe((/**
                 * @return {?}
                 */
                function () {
                    recompute(_this.vueInstance, stateName);
                }));
                this.subscriptionsMap.set(stateName, subscription);
            }
            return this.viewModelContext.uiState;
        };
        /**
         * 映射表单
         */
        /**
         * 映射表单
         * @return {?}
         */
        BindingStates.prototype.getForm = /**
         * 映射表单
         * @return {?}
         */
        function () {
            var _this = this;
            /** @type {?} */
            var stateName = 'form';
            if (this.subscriptionsMap.has(stateName) === false) {
                /** @type {?} */
                var subscription = this.viewModelContext.bindingData.changes.subscribe((/**
                 * @param {?} change
                 * @return {?}
                 */
                function (change) {
                    if (change && change.type === "ValueChanged") {
                        _this.viewModelContext.form.getFieldErrorByPath(change.path);
                    }
                    recompute(_this.vueInstance, stateName);
                }));
                /** @type {?} */
                var subscriptionV = this.viewModelContext.form.changes.subscribe((/**
                 * @param {?} data
                 * @return {?}
                 */
                function (data) {
                    if (data && data.type === 'validateFieldsFinished') {
                        recompute(_this.vueInstance, stateName);
                    }
                    else if (data && data.type === 'singleHanded') {
                        /** @type {?} */
                        var index = _this.vueInstance.$children.findIndex((/**
                         * @param {?} item
                         * @return {?}
                         */
                        function (item) { return item.$attrs.id === "form_" + data.value; }));
                        _this.vueInstance.$children[index].$forceUpdate();
                    }
                }));
                this.subscriptionsMap.set(stateName, subscription);
                this.subscriptionsMap.set(stateName + 'V', subscriptionV);
            }
            return this.vueInstance.$getViewModel().form;
        };
        /**
         * 映射状态机
         */
        /**
         * 映射状态机
         * @return {?}
         */
        BindingStates.prototype.getStateMachine = /**
         * 映射状态机
         * @return {?}
         */
        function () {
            var _this = this;
            /** @type {?} */
            var stateName = 'stateMachine';
            if (this.subscriptionsMap.has(stateName) === false) {
                /** @type {?} */
                var subscription = this.viewModelContext.stateMachine.stateChange.subscribe((/**
                 * @param {?} change
                 * @return {?}
                 */
                function (change) {
                    recompute(_this.vueInstance, stateName);
                }));
                this.subscriptionsMap.set(stateName, subscription);
            }
            return this.viewModelContext.stateMachine.renderStates;
        };
        return BindingStates;
    }());
    if (false) {
        /**
         * Vue实例
         * @type {?}
         * @private
         */
        BindingStates.prototype.vueInstance;
        /**
         * Vue实例的ViewModelContext
         * @type {?}
         * @private
         */
        BindingStates.prototype.viewModelContext;
        /**
         * 订阅集合
         * @type {?}
         * @private
         */
        BindingStates.prototype.subscriptionsMap;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component-extenders/component-extender.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 组件扩展器
     * @record
     */
    function ComponentExtender() { }
    if (false) {
        /**
         * 扩展组件
         * @return {?}
         */
        ComponentExtender.prototype.extend = function () { };
        /**
         * 清空绑定状态
         * @return {?}
         */
        ComponentExtender.prototype.dispose = function () { };
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component-extenders/listview-extender.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 列表组件扩展
     */
    var /**
     * 列表组件扩展
     */
    ListviewExtender = /** @class */ (function () {
        function ListviewExtender(listviewInstance, viewModelContext) {
            this.listviewInstance = listviewInstance;
            this.viewModelContext = viewModelContext;
        }
        /**
         * 扩展组件
         */
        /**
         * 扩展组件
         * @return {?}
         */
        ListviewExtender.prototype.extend = /**
         * 扩展组件
         * @return {?}
         */
        function () {
            var _this = this;
            /** @type {?} */
            var self = this;
            this.viewModelContext.appContext.eventBus.subscribe(this.viewModelContext.id + "-listview-extend", (/**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                if (value.type === 'listviewFinished') {
                    _this.viewModelContext.uiState.setPropertyValue('listaviewState', __assign({}, _this.viewModelContext.uiState['listaviewState'], { loadMoreFinished: true }));
                }
                if (value.type === 'listviewUnFinished') {
                    _this.viewModelContext.uiState.setPropertyValue('listaviewState', __assign({}, _this.viewModelContext.uiState['listaviewState'], { loadMoreFinished: false }));
                }
                if (value.type === 'loadDataFinished') {
                    _this.viewModelContext.uiState.setPropertyValue('listaviewState', __assign({}, _this.viewModelContext.uiState['listaviewState'], { refreshLoading: false, loadMoreLoading: false }));
                }
            }));
            // 当前行
            this.listviewInstance.$on('selectChange', (/**
             * @param {?} selectedData
             * @return {?}
             */
            function (selectedData) {
                /** @type {?} */
                var selectedId = selectedData.data.id;
                self.viewModelContext.bindingData.getList().setCurrentId(selectedId, true, true);
            }));
            // 批量选择
            this.listviewInstance.$on('itemClick', (/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var selectedItems = self.listviewInstance.getSelectDataList();
                /** @type {?} */
                var selectedIds = [];
                if (selectedItems && selectedItems.length > 0) {
                    selectedIds = selectedItems.map((/**
                     * @param {?} selectedItem
                     * @return {?}
                     */
                    function (selectedItem) {
                        return selectedItem.id;
                    }));
                }
                self.viewModelContext.uiState.setPropertyValue('selectedIds', selectedIds);
            }));
            // 监听下拉刷新组件
            this.listviewInstance.$on('refresh', (/**
             * @return {?}
             */
            function () {
                self.viewModelContext.repository.setPaginationInfo({
                    pageIndex: 1
                });
                self.viewModelContext.uiState.setPropertyValue('listaviewState', __assign({}, self.viewModelContext.uiState['listaviewState'], { refreshLoading: true, loadMoreLoading: true }));
                self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
            }));
            // 监听是否走查询
            this.listviewInstance.$on('load', (/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var loading = self.viewModelContext.uiState['listaviewState']['loadMoreLoading'];
                if (!loading) {
                    self.viewModelContext.uiState.setPropertyValue('listaviewState', __assign({}, self.viewModelContext.uiState['listaviewState'], { loadMoreLoading: true }));
                    /** @type {?} */
                    var pageIndex = self.viewModelContext.repository.paginationInfo && self.viewModelContext.repository.paginationInfo.pageIndex;
                    if (!pageIndex) {
                        pageIndex = 0;
                    }
                    self.viewModelContext.repository.setPaginationInfo({
                        pageIndex: pageIndex += 1
                    });
                    self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
                }
            }));
        };
        /**
         * 清理状态
         */
        /**
         * 清理状态
         * @return {?}
         */
        ListviewExtender.prototype.dispose = /**
         * 清理状态
         * @return {?}
         */
        function () {
            this.listviewInstance.$off('selectChange');
            this.listviewInstance.$off('itemClick');
            this.listviewInstance.$off('load');
            this.viewModelContext.appContext.eventBus.unsubscribe(this.viewModelContext.id + "-listview-extend");
        };
        return ListviewExtender;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        ListviewExtender.prototype.listviewInstance;
        /**
         * @type {?}
         * @private
         */
        ListviewExtender.prototype.viewModelContext;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component-extenders/lookup-extender.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 帮助控件扩展
     */
    var /**
     * 帮助控件扩展
     */
    LookupExtender = /** @class */ (function () {
        function LookupExtender(listviewInstance, viewModelContext) {
            this.lookupInstance = listviewInstance;
            this.viewModelContext = viewModelContext;
        }
        /**
         * @return {?}
         */
        LookupExtender.prototype.extend = /**
         * @return {?}
         */
        function () {
        };
        /**
         * @return {?}
         */
        LookupExtender.prototype.dispose = /**
         * @return {?}
         */
        function () {
            console.log('Lookup disposed...');
        };
        return LookupExtender;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        LookupExtender.prototype.lookupInstance;
        /**
         * @type {?}
         * @private
         */
        LookupExtender.prototype.viewModelContext;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component-extenders/attachment-extender.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 列表组件扩展
     */
    var /**
     * 列表组件扩展
     */
    AttachmentExtender = /** @class */ (function () {
        function AttachmentExtender(attachmentInstance, viewModelContext) {
            this.attachmentInstance = attachmentInstance;
            this.viewModelContext = viewModelContext;
        }
        /**
         * 扩展组件
         */
        /**
         * 扩展组件
         * @return {?}
         */
        AttachmentExtender.prototype.extend = /**
         * 扩展组件
         * @return {?}
         */
        function () {
            /** @type {?} */
            var self = this;
            // 当前行
            this.attachmentInstance.$on('beforeSelect', (/**
             * @param {?} selectedData
             * @return {?}
             */
            function (selectedData) {
                /** @type {?} */
                var selectedId = selectedData.id;
                self.viewModelContext.bindingData.getList().setCurrentId(selectedId, true, false);
            }));
        };
        /**
         * 清理状态
         */
        /**
         * 清理状态
         * @return {?}
         */
        AttachmentExtender.prototype.dispose = /**
         * 清理状态
         * @return {?}
         */
        function () {
            this.attachmentInstance.$off('beforeSelect');
        };
        return AttachmentExtender;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        AttachmentExtender.prototype.attachmentInstance;
        /**
         * @type {?}
         * @private
         */
        AttachmentExtender.prototype.viewModelContext;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component-extenders/filter-extender.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 列表组件扩展
     */
    var /**
     * 列表组件扩展
     */
    FilterExtender = /** @class */ (function () {
        function FilterExtender(filterInstance, viewModelContext) {
            this.filterInstance = filterInstance;
            this.viewModelContext = viewModelContext;
        }
        /**
         * 扩展组件
         */
        /**
         * 扩展组件
         * @return {?}
         */
        FilterExtender.prototype.extend = /**
         * 扩展组件
         * @return {?}
         */
        function () {
            /** @type {?} */
            var self = this;
            // 清空uistate的筛选条件
            this.filterInstance.$on('clear', (/**
             * @return {?}
             */
            function () {
                self.viewModelContext.uiState.setPropertyValue('filterState', { filters: null });
            }));
            // 给uistate的赋值筛选条件
            // this.filterInstance.$on('change', (data) => {
            //   //self.viewModelContext.uiState.setPropertyValue('filterState', { filters: data.FilterConditions });
            // });
            // 给uistate的赋值筛选条件
            this.filterInstance.$on('search', (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                self.viewModelContext.repository.setPaginationInfo({
                    pageIndex: 1
                });
                self.viewModelContext.uiState.setPropertyValue('filterState', { filters: data.FilterConditions });
                self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
            }));
            // 监听确定时间出发查询
            this.filterInstance.$on('confirm', (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                self.viewModelContext.repository.setPaginationInfo({
                    pageIndex: 1
                });
                self.viewModelContext.uiState.setPropertyValue('filterState', { filters: data.FilterConditions });
                self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
            }));
        };
        /**
         * 清理状态
         */
        /**
         * 清理状态
         * @return {?}
         */
        FilterExtender.prototype.dispose = /**
         * 清理状态
         * @return {?}
         */
        function () {
            this.filterInstance.$off('clear');
            this.filterInstance.$off('change');
            this.filterInstance.$off('confirm');
        };
        return FilterExtender;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        FilterExtender.prototype.filterInstance;
        /**
         * @type {?}
         * @private
         */
        FilterExtender.prototype.viewModelContext;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component-extenders/tab-extender.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 列表组件扩展
     */
    var /**
     * 列表组件扩展
     */
    TabExtender = /** @class */ (function () {
        function TabExtender(tabInstance, viewModelContext) {
            this.tabInstance = tabInstance;
            this.viewModelContext = viewModelContext;
            this.viewModelContext.uiState.setPropertyValue('tabState', { currentTab: tabInstance.currentName });
        }
        /**
         * 扩展组件
         */
        /**
         * 扩展组件
         * @return {?}
         */
        TabExtender.prototype.extend = /**
         * 扩展组件
         * @return {?}
         */
        function () {
            /** @type {?} */
            var self = this;
            // 给uistate的赋值当前tab页
            this.tabInstance.$on('change', (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                window['MOBILE_ORIGIN_BACK'] && window['MOBILE_ORIGIN_BACK'].reflushOriginGoback();
                if (data.name !== self.viewModelContext.uiState['tabState']['currentTab']) {
                    self.viewModelContext.uiState.setPropertyValue('tabState', { currentTab: data.name });
                    self.viewModelContext.repository.setPaginationInfo({ pageIndex: 1 });
                    self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
                }
            }));
        };
        /**
         * 清理状态
         */
        /**
         * 清理状态
         * @return {?}
         */
        TabExtender.prototype.dispose = /**
         * 清理状态
         * @return {?}
         */
        function () {
            this.tabInstance.$off('change');
        };
        return TabExtender;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        TabExtender.prototype.tabInstance;
        /**
         * @type {?}
         * @private
         */
        TabExtender.prototype.viewModelContext;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component-extenders/search-extender.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 列表组件扩展
     */
    var /**
     * 列表组件扩展
     */
    SearchExtender = /** @class */ (function () {
        function SearchExtender(searchInstance, viewModelContext) {
            this.searchInstance = searchInstance;
            this.viewModelContext = viewModelContext;
        }
        /**
         * 扩展组件
         */
        /**
         * 扩展组件
         * @return {?}
         */
        SearchExtender.prototype.extend = /**
         * 扩展组件
         * @return {?}
         */
        function () {
            /** @type {?} */
            var self = this;
            // 监听搜索组件
            this.searchInstance.$on('cancel', (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                //  清空走查询
                self.viewModelContext.repository.setPaginationInfo({
                    pageIndex: 1
                });
                self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
            }));
            this.searchInstance.$on('search', (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                // 如果是确定类型，即走查询
                self.viewModelContext.repository.setPaginationInfo({
                    pageIndex: 1
                });
                self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
            }));
        };
        /**
         * 清理状态
         */
        /**
         * 清理状态
         * @return {?}
         */
        SearchExtender.prototype.dispose = /**
         * 清理状态
         * @return {?}
         */
        function () {
            this.searchInstance.$off('cancel');
            this.searchInstance.$off('search');
        };
        return SearchExtender;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        SearchExtender.prototype.searchInstance;
        /**
         * @type {?}
         * @private
         */
        SearchExtender.prototype.viewModelContext;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component-extenders/checkbox-extender.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 列表组件扩展
     */
    var /**
     * 列表组件扩展
     */
    CheckboxExtender = /** @class */ (function () {
        function CheckboxExtender(checkboxInstance, viewModelContext) {
            this.checkboxInstance = checkboxInstance;
            this.viewModelContext = viewModelContext;
        }
        /**
         * 扩展组件
         */
        /**
         * 扩展组件
         * @return {?}
         */
        CheckboxExtender.prototype.extend = /**
         * 扩展组件
         * @return {?}
         */
        function () {
            /** @type {?} */
            var self = this;
            // 当按钮被加载调用父组件让其出现多选框
            this.checkboxInstance.$parent.listCheckStart();
            this.checkboxInstance.$on('click', (/**
             * @return {?}
             */
            function () {
                console.log("click");
                // 监听全选
                if (self.checkboxInstance.$parent.checkboxFlag) {
                    self.checkboxInstance.$parent.listCheckAllCancel();
                }
                else {
                    self.checkboxInstance.$parent.listCheckAll();
                }
                self.checkboxInstance.$parent.checkboxFlag = !self.checkboxInstance.$parent.checkboxFlag;
                /** @type {?} */
                var selectedItems = self.checkboxInstance.$parent.getSelectDataList();
                /** @type {?} */
                var selectedIds = [];
                if (selectedItems && selectedItems.length > 0) {
                    selectedIds = selectedItems.map((/**
                     * @param {?} selectedItem
                     * @return {?}
                     */
                    function (selectedItem) {
                        return selectedItem.id;
                    }));
                }
                self.viewModelContext.uiState.setPropertyValue('selectedIds', selectedIds);
            }));
        };
        /**
         * 清理状态
         */
        /**
         * 清理状态
         * @return {?}
         */
        CheckboxExtender.prototype.dispose = /**
         * 清理状态
         * @return {?}
         */
        function () {
            this.checkboxInstance.$off('click');
        };
        return CheckboxExtender;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        CheckboxExtender.prototype.checkboxInstance;
        /**
         * @type {?}
         * @private
         */
        CheckboxExtender.prototype.viewModelContext;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component-extenders/map-extender.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 列表组件扩展
     */
    var /**
     * 列表组件扩展
     */
    MapExtender = /** @class */ (function () {
        function MapExtender(mapInstance, viewModelContext) {
            this.mapInstance = mapInstance;
            this.viewModelContext = viewModelContext;
        }
        /**
         * 扩展组件
         */
        /**
         * 扩展组件
         * @return {?}
         */
        MapExtender.prototype.extend = /**
         * 扩展组件
         * @return {?}
         */
        function () {
            var _this = this;
            /** @type {?} */
            var self = this;
            // 当前行
            this.mapInstance.$on('markMap', (/**
             * @param {?} selectedData
             * @return {?}
             */
            function (selectedData) {
                // 获取经纬度
                /** @type {?} */
                var lnglat = selectedData.lnglat;
                /** @type {?} */
                var lng = lnglat.lng;
                /** @type {?} */
                var lat = lnglat.lat;
                // 获取经纬度路径
                /** @type {?} */
                var lngPath = _this.mapInstance.lngPath;
                /** @type {?} */
                var latPath = _this.mapInstance.latPath;
                // this.mapInstance.center = [lng, lat];
                // 赋值经纬度
                _this.viewModelContext.bindingData.setValue(_this.getBindingPathArray().concat(lngPath.split('/')), lng, true, true);
                _this.viewModelContext.bindingData.setValue(_this.getBindingPathArray().concat(latPath.split('/')), lat, true, true);
            }));
        };
        /**
         * 清理状态
         */
        /**
         * 清理状态
         * @return {?}
         */
        MapExtender.prototype.dispose = /**
         * 清理状态
         * @return {?}
         */
        function () {
            this.mapInstance.$off('markMap');
        };
        /**
         * @private
         * @return {?}
         */
        MapExtender.prototype.getBindingPathArray = /**
         * @private
         * @return {?}
         */
        function () {
            /** @type {?} */
            var path = this.viewModelContext.bindingData.bindingPath;
            if (path) {
                return path.split('/').filter((/**
                 * @param {?} n
                 * @return {?}
                 */
                function (n) { return n !== ''; }));
            }
            return [];
        };
        return MapExtender;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        MapExtender.prototype.mapInstance;
        /**
         * @type {?}
         * @private
         */
        MapExtender.prototype.viewModelContext;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component-extenders/component-extender-factory.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 组件适配器工厂
     */
    var /**
     * 组件适配器工厂
     */
    ComponentExtenderFactory = /** @class */ (function () {
        function ComponentExtenderFactory(viewModelContext) {
            this.viewModelContext = viewModelContext;
        }
        /**
         * 创建适配器
         */
        /**
         * 创建适配器
         * @param {?} componentInstance
         * @param {?} componentType
         * @return {?}
         */
        ComponentExtenderFactory.prototype.create = /**
         * 创建适配器
         * @param {?} componentInstance
         * @param {?} componentType
         * @return {?}
         */
        function (componentInstance, componentType) {
            /** @type {?} */
            var componentExtenderInstance;
            switch (componentType) {
                case 'fm-listview':
                    componentExtenderInstance = new ListviewExtender(componentInstance, this.viewModelContext);
                    break;
                case 'Lookup':
                    componentExtenderInstance = new LookupExtender(componentInstance, this.viewModelContext);
                    break;
                case 'fm-filter':
                    componentExtenderInstance = new FilterExtender(componentInstance, this.viewModelContext);
                    break;
                case 'fm-tabs':
                    componentExtenderInstance = new TabExtender(componentInstance, this.viewModelContext);
                    break;
                case 'van-search':
                    componentExtenderInstance = new SearchExtender(componentInstance, this.viewModelContext);
                    break;
                case 'Checkbox':
                    componentExtenderInstance = new CheckboxExtender(componentInstance, this.viewModelContext);
                    break;
                case 'fm-attachment':
                    componentExtenderInstance = new AttachmentExtender(componentInstance, this.viewModelContext);
                    break;
                case 'Map':
                    componentExtenderInstance = new MapExtender(componentInstance, this.viewModelContext);
                    break;
                default:
                    componentExtenderInstance = null;
            }
            return componentExtenderInstance;
        };
        return ComponentExtenderFactory;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        ComponentExtenderFactory.prototype.viewModelContext;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component-extenders/component-extender-manager.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 扩展管理器
     */
    var /**
     * 扩展管理器
     */
    ComponentExtenderManager = /** @class */ (function () {
        /**
         * 构造函数
         */
        function ComponentExtenderManager(viewModelContext) {
            this.viewModelContext = viewModelContext;
            this.extenderMap = new Map();
            this.extenderFactory = new ComponentExtenderFactory(viewModelContext);
        }
        /**
         * 获取扩展
         */
        /**
         * 获取扩展
         * @param {?} componentInstance
         * @param {?} componentType
         * @return {?}
         */
        ComponentExtenderManager.prototype.getExtender = /**
         * 获取扩展
         * @param {?} componentInstance
         * @param {?} componentType
         * @return {?}
         */
        function (componentInstance, componentType) {
            if (this.extenderMap.has(componentInstance)) {
                return this.extenderMap.get(componentInstance);
            }
            /** @type {?} */
            var extender = this.extenderFactory.create(componentInstance, componentType);
            this.extenderMap.set(componentInstance, extender);
            return extender;
        };
        return ComponentExtenderManager;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        ComponentExtenderManager.prototype.viewModelContext;
        /**
         * 组件扩展器集合
         * @type {?}
         * @private
         */
        ComponentExtenderManager.prototype.extenderMap;
        /**
         * 扩展器工厂
         * @type {?}
         * @private
         */
        ComponentExtenderManager.prototype.extenderFactory;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/component-extenders/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/utils/get-data-by-path.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var getDataByPath = (/**
     * @param {?} data
     * @param {?} fields
     * @return {?}
     */
    function (data, fields) {
        if (!data[fields[0]]) {
            return;
        }
        return fields.reduce((/**
         * @param {?} total
         * @param {?} next
         * @return {?}
         */
        function (total, next) {
            return total[next];
        }), data);
    });
    var ɵ0 = getDataByPath;

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/utils/attachment-util.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var mergeString = (/**
     * @param {?} str
     * @param {?} str1
     * @return {?}
     */
    function (str, str1) {
        /** @type {?} */
        var min = str.length < str1.length ? str.length : str1.length;
        /** @type {?} */
        var res = "";
        /** @type {?} */
        var i = 0;
        for (i = 0; i < min; i++) {
            res += str.substring(i, i + 1);
            res += str1.substring(i, i + 1);
        }
        res += str.length < str1.length ? str1.substring(i) : str.substring(i);
        return res;
    });
    var ɵ0$1 = mergeString;
    /** @type {?} */
    var getToken = (/**
     * @param {?} str
     * @return {?}
     */
    function (str) {
        /** @type {?} */
        var array = [1, 2, 3];
        if (str.length > 24) {
            str = str.substring(0, 24);
        }
        /** @type {?} */
        var length = Math.floor(str.length / 3);
        /** @type {?} */
        var str1 = str.substring(0, length);
        /** @type {?} */
        var str2 = str.substring(length, 2 * length);
        /** @type {?} */
        var str3 = str.substring(2 * length, 3 * length);
        /** @type {?} */
        var randomArray = [];
        /** @type {?} */
        var random = Math.ceil(Math.random() * 3);
        randomArray.push(array[random - 1]);
        array.splice(random - 1, 1);
        random = Math.ceil(Math.random() * 2);
        randomArray.push(array[random - 1]);
        array.splice(random - 1, 1);
        randomArray.push(array[0]);
        /** @type {?} */
        var timeStamp = Math.round(new Date().getTime() / 1000) + "";
        return timeStamp.substring(0, 5) + "times" + randomArray[0] + ""
            + randomArray[1] + "" + randomArray[2] + switchStr(str1, str2, str3, randomArray[0])
            + switchStr(str1, str2, str3, randomArray[1]) +
            switchStr(str1, str2, str3, randomArray[2]) + "times" + timeStamp.substring(5);
    });
    var ɵ1 = getToken;
    /** @type {?} */
    var switchStr = (/**
     * @param {?} str1
     * @param {?} str2
     * @param {?} str3
     * @param {?} i
     * @return {?}
     */
    function (str1, str2, str3, i) {
        switch (i) {
            case 1: return str1;
            case 2: return str2;
            case 3: return str3;
        }
    });
    var ɵ2 = switchStr;
    /** @type {?} */
    var officeUrl = {
        doc: '/platform/common/web/assets/fm-attachment/word.png',
        docx: '/platform/common/web/assets/fm-attachment/word.png',
        xls: '/platform/common/web/assets/fm-attachment/excel.png',
        xlsx: '/platform/common/web/assets/fm-attachment/excel.png',
        ppt: '/platform/common/web/assets/fm-attachment/ppt.png',
        pptx: '/platform/common/web/assets/fm-attachment/ppt.png',
        pdf: '/platform/common/web/assets/fm-attachment/pdf.png',
        txt: '/platform/common/web/assets/fm-attachment/txt.png',
        jpg: '/platform/common/web/assets/fm-attachment/img.png',
        jpeg: '/platform/common/web/assets/fm-attachment/img.png',
        png: '/platform/common/web/assets/fm-attachment/img.png'
    };
    /** @type {?} */
    var getAttachmentUrl = (/**
     * @param {?} attachmentId
     * @param {?} rootId
     * @param {?} extensionName
     * @return {?}
     */
    function (attachmentId, rootId, extensionName) {
        /** @type {?} */
        var officeUrlKey = Object.keys(officeUrl).find((/**
         * @param {?} url
         * @return {?}
         */
        function (url) { return url === extensionName.toLowerCase(); }));
        if (!officeUrlKey) {
            return "/api/runtime/dfs/v1.0/doc/filecontent?metadataid=" + attachmentId + "&rootid=" + rootId + "&token=" + getToken(mergeString(attachmentId, rootId));
        }
        return officeUrl[extensionName.toLowerCase()];
    });
    var ɵ3 = getAttachmentUrl;

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/utils/event-util.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var EventUtil = /** @class */ (function () {
        function EventUtil() {
        }
        /**
         * @param {?} str
         * @return {?}
         */
        EventUtil.toHandlerKey = /**
         * @param {?} str
         * @return {?}
         */
        function (str) {
            if (str) {
                return "on" + this.capitalize(str);
            }
            else {
                return '';
            }
        };
        /**
         * @param {?} str
         * @return {?}
         */
        EventUtil.camelize = /**
         * @param {?} str
         * @return {?}
         */
        function (str) {
            /** @type {?} */
            var camelizeRE = /-(\w)/g;
            return str.replace(camelizeRE, (/**
             * @param {?} _
             * @param {?} c
             * @return {?}
             */
            function (_, c) { return (c ? c.toUpperCase() : ''); }));
        };
        /**
         * @param {?} str
         * @return {?}
         */
        EventUtil.capitalize = /**
         * @param {?} str
         * @return {?}
         */
        function (str) {
            return str.charAt(0).toUpperCase() + str.slice(1);
        };
        return EventUtil;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/utils/uuid.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var uuid = (/**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var d = new Date().getTime();
        /** @type {?} */
        var uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (/**
         * @param {?} c
         * @return {?}
         */
        function (c) {
            /** @type {?} */
            var r = (d + Math.random() * 16) % 16 | 0;
            d = Math.floor(d / 16);
            return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
        }));
        return uuid;
    });

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/utils/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/utils/format-uitl.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var formatUtils = {
        /**
         * 枚举格式化
         */
        enum: (/**
         * @param {?} value
         * @param {?} options
         * @return {?}
         */
        function (value, options) {
            return mobileDevkit.EnumUtil.format(value, options);
        }),
        /**
         * 日期格式化
         */
        date: (/**
         * @param {?} value
         * @param {?} options
         * @return {?}
         */
        function (value, options) {
            return mobileDevkit.DateUtil.format(value, options);
        }),
        /**
         * 日期操作
         */
        dateOperation: (/**
         * @param {?} value
         * @param {?} options
         * @return {?}
         */
        function (value, options) {
            return mobileDevkit.DateUtil.dateOperation(value, options);
        }),
        /**
         * 数字格式化
         */
        number: (/**
         * @param {?} value
         * @param {?} options
         * @return {?}
         */
        function (value, options) {
            return mobileDevkit.NumberUtil.format(value, options);
        }),
        /**
         * 字符串格式化
         *
         */
        string: (/**
         * @param {?} value
         * @param {?} options
         * @return {?}
         */
        function (value, options) {
            return mobileDevkit.StringUtil.format(value, options);
        }),
        /**
         * 布尔值格式化
         */
        bool: (/**
         * @param {?} value
         * @param {?} options
         * @return {?}
         */
        function (value, options) {
            return mobileDevkit.BoolUtil.format(value, options);
        }),
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/devkit.plugin.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ɵ0$2 = /**
     * @param {?} app
     * @param {?} options
     * @return {?}
     */
    function (app, options) {
        /**
         * 注册事件
         * @todo：暂时强制绑定，暂时无法根据特性灵活区分
         */
        app.config.globalProperties.$on = (/**
         * @param {?} eventName
         * @param {?} eventHandler
         * @return {?}
         */
        function (eventName, eventHandler) {
            /** @type {?} */
            var vnodeProps = this._.vnode.props;
            /** @type {?} */
            var eventHandlerKey = EventUtil.toHandlerKey(eventName);
            /** @type {?} */
            var oldEventHandler = vnodeProps[eventHandlerKey];
            if (oldEventHandler) {
                vnodeProps[eventHandlerKey] = (/**
                 * @param {?} $event
                 * @return {?}
                 */
                function ($event) {
                    eventHandler($event);
                    oldEventHandler($event);
                });
            }
            else {
                vnodeProps[eventHandlerKey] = eventHandler;
            }
        });
        /**
         * 清空事件注册
         * @todo：暂时强制清空
         */
        app.config.globalProperties.$off = (/**
         * @param {?} eventName
         * @return {?}
         */
        function (eventName) {
            /** @type {?} */
            var vnodeProps = this._.vnode.props;
            /** @type {?} */
            var eventHandlerKey = EventUtil.toHandlerKey(eventName);
            vnodeProps[eventHandlerKey] = null;
        });
        /**
         * 组件的App实例
         */
        app.config.globalProperties.$app = null;
        /**
         * 组件的ViewModel实例
         */
        app.config.globalProperties.$viewModel = null;
        /**
         * 绑定状态
         */
        app.config.globalProperties.$bindingStates = null;
        /**
         * 绑定状态
         */
        app.config.globalProperties.$componentExtenderManager = null;
        /**
         * 组件App工厂方法
         */
        app.config.globalProperties.$createApp = (/**
         * @param {?} appOptions
         * @return {?}
         */
        function (appOptions) {
            if (this.$app) {
                throw new Error('App已经存在');
            }
            /** @type {?} */
            var historyBackService = new mobileCommandServices.HistoryBackService();
            historyBackService.init();
            historyBackService.proxyBack();
            this.$app = new mobileDevkit.App(appOptions);
            /**
             * @param {?} event
             * @return {?}
             */
            function receiveMessage(event) {
                /** @type {?} */
                var commandOptions = this.$app.context.injector.get('DECLARATIONS_TOKEN');
                /** @type {?} */
                var messageData = JSON.parse(event.data);
                /** @type {?} */
                var commandName = messageData.data.command;
                /** @type {?} */
                var command = commandOptions.commands[commandName];
                /** @type {?} */
                var origin = event.origin;
                /** @type {?} */
                var source = event.source;
                /** @type {?} */
                var viewModelContext = this.$app.context.viewModelContextManager.getContextById(command.viewModelId);
                if (!!viewModelContext) {
                    /** @type {?} */
                    var methodResult$ = viewModelContext.viewModel[command.command]();
                    methodResult$.pipe((/**
                     * @return {?}
                     */
                    function () { }), (/**
                     * @return {?}
                     */
                    function () { }), (/**
                     * @return {?}
                     */
                    function () {
                        window.parent && window.parent.postMessage('success', event.origin);
                    }));
                }
            }
            if (!window['IsDddedMessageEventListener']) {
                window.addEventListener("message", receiveMessage.bind(this), false);
                window['IsDddedMessageEventListener'] = true;
            }
        });
        /**
         * 获取AppContext
         */
        app.config.globalProperties.$getApp = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var app = this.$app;
            /** @type {?} */
            var currentCmp = this;
            while (!app && currentCmp.$parent) {
                currentCmp = currentCmp.$parent;
                app = currentCmp.$app;
            }
            return app;
        });
        /**
         * 组件ViewModel工厂方法
         */
        app.config.globalProperties.$createViewModel = (/**
         * @param {?} viewModelOptions
         * @return {?}
         */
        function (viewModelOptions) {
            this.$app = this.$getApp();
            if (!this.$app) {
                throw new Error('请先创建App');
            }
            // 如果viewModel已经存在，则使用已有的
            /** @type {?} */
            var viewModelId = viewModelOptions.id;
            if (viewModelId) {
                /** @type {?} */
                var existedViewModelContext = this.$app.context.viewModelContextManager.getContextById(viewModelId);
                if (existedViewModelContext) {
                    this.$viewModel = existedViewModelContext.viewModel;
                }
            }
            // 如果viewModel不存在，则创建新的viewModel，并建立父子关系
            if (!this.$viewModel) {
                /** @type {?} */
                var parentViewModel = this.$getViewModel();
                if (!viewModelOptions.parent) {
                    viewModelOptions.parent = parentViewModel;
                }
                this.$viewModel = this.$app.createViewModel(viewModelOptions);
            }
            // 设置绑定状态
            this.$bindingStates = new BindingStates(this);
            // 映射路由状态
            /** @type {?} */
            var routerService = this.$viewModel.injector.get(mobileCommandServices.RouterService);
            routerService.mappingToUIState();
            // 根据路由参数清理卡片缓存数据
            if (this.$viewModel && this.$viewModel.uiState && this.$viewModel.uiState.routerState && this.$viewModel.uiState.routerState.queryParams && this.$viewModel.uiState.routerState.queryParams.clearCachedDataForCard) {
                //清一下bingdingdate上的东西
                this.$viewModel.context.bindingData.getList().currentId = null;
            }
            // 根据路由参数清理列表缓存数据
            if (this.$viewModel && this.$viewModel.uiState && this.$viewModel.uiState.routerState && this.$viewModel.uiState.routerState.queryParams && this.$viewModel.uiState.routerState.queryParams.clearCachedDataForList) {
                //清一下bingdingdate上的东西
                this.$viewModel.repository.entityCollection.clear();
            }
        });
        /**
         * 获取ViewModel
         */
        app.config.globalProperties.$getViewModel = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var currentCmp = this;
            /** @type {?} */
            var viewModel = currentCmp.$viewModel;
            while (!viewModel && currentCmp.$parent) {
                currentCmp = currentCmp.$parent;
                viewModel = currentCmp.$viewModel;
            }
            return viewModel;
        });
        /**
         * 初始化组件扩展管理
         */
        app.config.globalProperties.$initComponentExtenderManager = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var viewModel = this.$getViewModel();
            this.$componentExtenderManager = new ComponentExtenderManager(viewModel.context);
        });
        /**
         * 获取组件扩展管理器
         */
        app.config.globalProperties.$getComponentExtenderManager = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var currentCmp = this;
            /** @type {?} */
            var extenderManager = currentCmp.$componentExtenderManager;
            while (!extenderManager && currentCmp.$parent) {
                currentCmp = currentCmp.$parent;
                extenderManager = currentCmp.$componentExtenderManager;
            }
            return extenderManager;
        });
        app.config.globalProperties.$bindComponentExtenders = (/**
         * @param {?} refs
         * @return {?}
         */
        function (refs) {
            var _this = this;
            if (!this.$refs) {
                return;
            }
            /** @type {?} */
            var componentInstances = Object.values(this.$refs);
            componentInstances.forEach((/**
             * @param {?} componentInstance
             * @return {?}
             */
            function (componentInstance) {
                /** @type {?} */
                var componentType = componentInstance._.type.name;
                _this.$bindComponentExtender(componentInstance, componentType);
            }));
        });
        /**
         * 绑定组件扩展器
         */
        app.config.globalProperties.$bindComponentExtender = (/**
         * @param {?} componentInstance
         * @param {?} componentType
         * @return {?}
         */
        function (componentInstance, componentType) {
            if (componentInstance.$__component_extended === true) {
                return;
            }
            /** @type {?} */
            var extenderManager = this.$getComponentExtenderManager();
            /** @type {?} */
            var extender = extenderManager.getExtender(componentInstance, componentType);
            if (extender) {
                extender.extend();
                componentInstance.$__component_extended === true;
            }
        });
        /**
         * 清理所有组件扩展器
         * @param refs
         */
        app.config.globalProperties.$unbindComponentExtenders = (/**
         * @param {?} refs
         * @return {?}
         */
        function (refs) {
            var _this = this;
            if (!this.$refs) {
                return;
            }
            /** @type {?} */
            var componentInstances = Object.values(this.$refs);
            componentInstances.forEach((/**
             * @param {?} componentInstance
             * @return {?}
             */
            function (componentInstance) {
                _this.$unbindComponentExtender(componentInstance);
            }));
        });
        /**
         * 清理组件扩展器状态
         */
        app.config.globalProperties.$unbindComponentExtender = (/**
         * @param {?} componentInstance
         * @return {?}
         */
        function (componentInstance) {
            /** @type {?} */
            var extenderManager = this.$getComponentExtenderManager();
            /** @type {?} */
            var extender = extenderManager.getExtender(componentInstance);
            if (extender) {
                extender.dispose();
            }
        });
        app.config.globalProperties.$initPage = (/**
         * @param {?} viewModelOptions
         * @return {?}
         */
        function (viewModelOptions) {
            this.$createViewModel(viewModelOptions);
            this.$initComponentExtenderManager();
            this.$viewModel.uiState.pageTitle && this.$changeDocumentTitle(this.$viewModel.uiState.pageTitle);
            this.$initUistate();
        });
        app.config.globalProperties.$changeDocumentTitle = (/**
         * @param {?} title
         * @return {?}
         */
        function (title) {
            document.title = title || document.title;
        });
        /**
         * VUE上格式化工具类
         */
        app.config.globalProperties.$formatUtils = formatUtils;
        /**
         * 初始化uistate上记录的状态
         */
        app.config.globalProperties.$initUistate = (/**
         * @return {?}
         */
        function () {
            this.$viewModel.uiState.listaviewState = { refreshLoading: false, loadMoreLoading: false, loadMoreFinished: false };
            this.$viewModel.uiState.searchState = { keyword: '' };
            this.$viewModel.uiState.filterState = { filters: '', filterData: '' };
            this.$viewModel.uiState.tabState = { currentTab: '' };
        });
    };
    /** @type {?} */
    var DevkitPlugin = {
        install: (ɵ0$2)
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/binding-states-mapper.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @return {?}
     */
    function mapEntityListData() {
        return recomputable((/**
         * @return {?}
         */
        function () {
            return this.$bindingStates.getEntityListData();
        }), 'entityListData');
    }
    /**
     * @return {?}
     */
    function mapCurrentEntityData() {
        return recomputable((/**
         * @return {?}
         */
        function () {
            return this.$bindingStates.getCurrentEntityData();
        }), 'currentEntityData');
    }
    /**
     * @return {?}
     */
    function mapUIState() {
        return recomputable((/**
         * @return {?}
         */
        function () {
            return this.$bindingStates.getUIState();
        }), 'uiState');
    }
    /**
     * @return {?}
     */
    function mapForm() {
        return recomputable((/**
         * @return {?}
         */
        function () {
            return this.$bindingStates.getForm();
        }), 'form');
    }
    /**
     * @return {?}
     */
    function mapStateMachine() {
        return recomputable((/**
         * @return {?}
         */
        function () {
            return this.$bindingStates.getStateMachine();
        }), 'stateMachine');
    }
    /**
     * @param {?=} extraStates
     * @return {?}
     */
    function mapBindingStates(extraStates) {
        /** @type {?} */
        var mappedBindingStates = {
            entityListData: mapEntityListData(),
            currentEntityData: mapCurrentEntityData(),
            uiState: mapUIState(),
            form: mapForm(),
            stateMachine: mapStateMachine()
        };
        /** @type {?} */
        var allMappedStates = Object.assign({}, mappedBindingStates, extraStates);
        return allMappedStates;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/binding-commands-mapper.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 映射命令方法
     * @param {?} methodNames
     * @return {?}
     */
    function mapCommandMethods(methodNames) {
        /** @type {?} */
        var mappedCommandMethods = {};
        methodNames.forEach((/**
         * @param {?} methodName
         * @return {?}
         */
        function (methodName) {
            mappedCommandMethods[methodName] = (/**
             * @param {?} eventParams
             * @return {?}
             */
            function (eventParams) {
                return this.$viewModel[methodName](eventParams);
            });
        }));
        return mappedCommandMethods;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/attachment/wechat-take-photo.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var WeChatTakePhoto = /** @class */ (function () {
        function WeChatTakePhoto() {
        }
        // public selectAttachments() { }
        // public selectAttachments() { }
        /**
         * @return {?}
         */
        WeChatTakePhoto.prototype.takePhotoAttachment = 
        // public selectAttachments() { }
        /**
         * @return {?}
         */
        function () {
            // 微信拍照
        };
        /**
         * @return {?}
         */
        WeChatTakePhoto.prototype.takeShortVideo = /**
         * @return {?}
         */
        function () {
        };
        /**
         * @return {?}
         */
        WeChatTakePhoto.prototype.selectFile = /**
         * @return {?}
         */
        function () {
        };
        /**
         * @return {?}
         */
        WeChatTakePhoto.prototype.downloadAttachment = /**
         * @return {?}
         */
        function () {
        };
        return WeChatTakePhoto;
    }());
    if (false) {
        /** @type {?} */
        WeChatTakePhoto.prototype.getSysInfo;
        /** @type {?} */
        WeChatTakePhoto.prototype.playShortVideo;
        /** @type {?} */
        WeChatTakePhoto.prototype.selectAttachments;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/attachment/attachment.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @abstract
     */
    var /**
     * @abstract
     */
    AttachmentService = /** @class */ (function () {
        function AttachmentService() {
            this.collectFile = new rxjs.Subject();
        }
        /**
         * 拍照
         * @param param 拍照参数
         */
        /**
         * 拍照
         * @param {?=} param 拍照参数
         * @return {?}
         */
        AttachmentService.prototype.takePhoto = /**
         * 拍照
         * @param {?=} param 拍照参数
         * @return {?}
         */
        function (param) {
            this.takePhotoIns.takePhotoAttachment(param);
        };
        /**
         * 选择文件
         * @param param 选择文件参数
         */
        /**
         * 选择文件
         * @param {?} param 选择文件参数
         * @return {?}
         */
        AttachmentService.prototype.selectAttachment = /**
         * 选择文件
         * @param {?} param 选择文件参数
         * @return {?}
         */
        function (param) {
            this.takePhotoIns.selectAttachments(param);
        };
        /**
         * 拍摄短视频
         * @param param 拍摄短视频参数
         */
        /**
         * 拍摄短视频
         * @param {?} param 拍摄短视频参数
         * @return {?}
         */
        AttachmentService.prototype.takeShortVideo = /**
         * 拍摄短视频
         * @param {?} param 拍摄短视频参数
         * @return {?}
         */
        function (param) {
            this.takePhotoIns.takeShortVideo(param);
        };
        /**
         * 播放短视频
         * @param param 播放短视频参数
         */
        /**
         * 播放短视频
         * @param {?} param 播放短视频参数
         * @return {?}
         */
        AttachmentService.prototype.playShortVideo = /**
         * 播放短视频
         * @param {?} param 播放短视频参数
         * @return {?}
         */
        function (param) {
            this.takePhotoIns.playShortVideo(param);
        };
        /**
         * @param {?} param
         * @return {?}
         */
        AttachmentService.prototype.selectFile = /**
         * @param {?} param
         * @return {?}
         */
        function (param) {
            this.takePhotoIns.selectFile(param);
        };
        /**
         * @param {?} param
         * @return {?}
         */
        AttachmentService.prototype.downloadAttachment = /**
         * @param {?} param
         * @return {?}
         */
        function (param) {
            this.takePhotoIns.downloadAttachment(param);
        };
        /**
         * @return {?}
         */
        AttachmentService.prototype.getSysInfo = /**
         * @return {?}
         */
        function () {
            return this.takePhotoIns.getSysInfo();
        };
        return AttachmentService;
    }());
    if (false) {
        /** @type {?} */
        AttachmentService.prototype.takePhotoIns;
        /** @type {?} */
        AttachmentService.prototype.collectFile;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/attachment/cloud-plus-take-photo.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var CloudPlusTakePhoto = /** @class */ (function () {
        function CloudPlusTakePhoto(collectFile) {
            this.collectFile = collectFile;
            this.jsBrigdeService = new mobileCommandServices.JsBridgeService();
        }
        /**
         * @param {?} param
         * @return {?}
         */
        CloudPlusTakePhoto.prototype.takeShortVideo = /**
         * @param {?} param
         * @return {?}
         */
        function (param) {
            var _this = this;
            this.jsBrigdeService.takeVideo(param).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                /** @type {?} */
                var data = _this.processVideo(result);
                _this.collectFile.next(data);
            }));
        };
        /**
         * @param {?} param
         * @return {?}
         */
        CloudPlusTakePhoto.prototype.takePhotoAttachment = /**
         * @param {?} param
         * @return {?}
         */
        function (param) {
            var _this = this;
            this.jsBrigdeService.takePhoto(param).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                /** @type {?} */
                var data = _this.processPic(result);
                _this.collectFile.next(data);
            }));
        };
        /**
         * @param {?} param
         * @return {?}
         */
        CloudPlusTakePhoto.prototype.selectAttachments = /**
         * @param {?} param
         * @return {?}
         */
        function (param) {
            var _this = this;
            this.jsBrigdeService.selectPictureFromAlbum(param).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                /** @type {?} */
                var data = _this.processPic(result);
                _this.collectFile.next(data);
            }));
        };
        /**
         * @param {?} param
         * @return {?}
         */
        CloudPlusTakePhoto.prototype.selectFile = /**
         * @param {?} param
         * @return {?}
         */
        function (param) {
            var _this = this;
            this.jsBrigdeService.selectFile(param).subscribe((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                /** @type {?} */
                var data = _this.processFile(res);
                _this.collectFile.next(data);
            }));
        };
        /**
         * @param {?} options
         * @return {?}
         */
        CloudPlusTakePhoto.prototype.downloadAttachment = /**
         * @param {?} options
         * @return {?}
         */
        function (options) {
            return this.jsBrigdeService.downloadFile(options);
        };
        /**
         * @return {?}
         */
        CloudPlusTakePhoto.prototype.getSysInfo = /**
         * @return {?}
         */
        function () {
            return this.jsBrigdeService.getSysInfo();
        };
        /**
         * 调起当前设备 播放短视频
         * @param path 视频路径
         */
        /**
         * 调起当前设备 播放短视频
         * @param {?} path 视频路径
         * @return {?}
         */
        CloudPlusTakePhoto.prototype.playShortVideo = /**
         * 调起当前设备 播放短视频
         * @param {?} path 视频路径
         * @return {?}
         */
        function (path) {
            var _this = this;
            window['iCityBridge'].ccworkPlayShortVideo({ path: path }, (/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                /** @type {?} */
                var result = res.result;
                /** @type {?} */
                var data = _this.processVideo(result);
                _this.collectFile.next(data);
            }));
        };
        /**
         * 预处理图片
         */
        /**
         * 预处理图片
         * @param {?} result
         * @return {?}
         */
        CloudPlusTakePhoto.prototype.processPic = /**
         * 预处理图片
         * @param {?} result
         * @return {?}
         */
        function (result) {
            /** @type {?} */
            var imgHeader = 'data:image/jpeg;base64,';
            /** @type {?} */
            var imgType = 'image/png';
            if (typeof result === 'string') {
                // ios takephoto
                return {
                    file: new File([], "image_" + uuid().substring(0, 6) + ".png", { type: imgType }),
                    content: "" + imgHeader + result
                };
            }
            return result.map((/**
             * @param {?} dataItem
             * @return {?}
             */
            function (dataItem) {
                // andriod selectPic
                if (typeof dataItem === 'string') {
                    return {
                        file: new File([], "image_" + uuid().substring(0, 6) + ".png", { type: imgType }),
                        content: "" + imgHeader + dataItem
                    };
                }
                // ios selectPic
                return {
                    file: new File([], "image_" + uuid().substring(0, 6) + ".png", { type: imgType }),
                    content: "" + imgHeader + dataItem.content
                };
            }));
        };
        /**
         * @param {?} result
         * @return {?}
         */
        CloudPlusTakePhoto.prototype.processFile = /**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            /** @type {?} */
            var fileHeaders = {
                'jpg': 'data:image/jpeg;base64,',
                'png': 'data:image/jpeg;base64,',
                'doc': 'data:application/vnd.openxmlformats-officedocument.wordprocessingml.document;base64,',
                'docx': 'data:application/vnd.openxmlformats-officedocument.wordprocessingml.document;base64,',
                'xls': 'data:application/vnd.openxmlformats-officedocument.spreadsheetml.sheet;base64,',
                'xlsx': 'data:application/vnd.openxmlformats-officedocument.spreadsheetml.sheet;base64,',
                'ppt': 'data:application/vnd.ms-powerpoint;base64,',
                'pptx': 'data:application/vnd.ms-powerpoint;base64,',
                'txt': 'data:text/plain;base64,',
                'pdf': 'data:application/pdf;base64,',
                'mp4': 'data:video/mp4;base64,',
                'mov': 'data:video/mp4;base64,'
            };
            /** @type {?} */
            var fileTypes = {
                'jpg': 'image/png',
                'png': 'image/png',
                'doc': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
                'docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
                'xls': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
                'xlsx': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
                'ppt': 'application/vnd.ms-powerpoint',
                'pptx': 'application/vnd.ms-powerpoint',
                'txt': 'text/plain',
                'pdf': 'application/pdf',
                'mp4': 'video/mp4',
                'mov': 'video/mp4'
            };
            return result.map((/**
             * @param {?} dataItem
             * @return {?}
             */
            function (dataItem) {
                /** @type {?} */
                var suffix = dataItem.name.substring(dataItem.name.lastIndexOf('.') + 1);
                return {
                    file: new File([], dataItem.name, { type: fileTypes[suffix.toLowerCase()] }),
                    content: "" + fileHeaders[suffix.toLowerCase()] + dataItem.base64
                };
            }));
        };
        /**
         * 预处理视频
         * @param result
         */
        /**
         * 预处理视频
         * @param {?} result
         * @return {?}
         */
        CloudPlusTakePhoto.prototype.processVideo = /**
         * 预处理视频
         * @param {?} result
         * @return {?}
         */
        function (result) {
            /** @type {?} */
            var url = result.value;
            /** @type {?} */
            var base64value = result.base64value;
            /** @type {?} */
            var videoHeader = 'data:video/mp4;base64,';
            /** @type {?} */
            var videoType = 'video/mp4';
            /** @type {?} */
            var videoName = url.substring(url.lastIndexOf('/') + 1) + '.mp4';
            return {
                file: new File([], videoName, { type: videoType }),
                content: "" + videoHeader + base64value,
                type: videoType,
                url: url,
                name: url.substring(url.lastIndexOf('/') + 1) + '.mp4'
            };
        };
        return CloudPlusTakePhoto;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        CloudPlusTakePhoto.prototype.jsBrigdeService;
        /**
         * @type {?}
         * @private
         */
        CloudPlusTakePhoto.prototype.collectFile;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/attachment/attachment-children.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var WeChatAttachment = /** @class */ (function (_super) {
        __extends(WeChatAttachment, _super);
        function WeChatAttachment() {
            var _this = _super.call(this) || this;
            _this.takePhotoIns = new WeChatTakePhoto();
            return _this;
        }
        return WeChatAttachment;
    }(AttachmentService));
    var CloudPlusAttachment = /** @class */ (function (_super) {
        __extends(CloudPlusAttachment, _super);
        function CloudPlusAttachment() {
            var _this = _super.call(this) || this;
            _this.takePhotoIns = new CloudPlusTakePhoto(_this.collectFile);
            return _this;
        }
        return CloudPlusAttachment;
    }(AttachmentService));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/attachment/upload-attachment.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var UploadAttachmentService = /** @class */ (function () {
        function UploadAttachmentService(befRepository) {
            this.befRepository = befRepository;
        }
        /**
         * 缓存附件
         * @param originFiles
         * @param uploadInfo
         */
        /**
         * 缓存附件
         * @param {?} originFiles
         * @param {?} uploadInfo
         * @return {?}
         */
        UploadAttachmentService.prototype.upload = /**
         * 缓存附件
         * @param {?} originFiles
         * @param {?} uploadInfo
         * @return {?}
         */
        function (originFiles, uploadInfo) {
            /** @type {?} */
            var apiProxy = this.befRepository.apiProxy;
            /** @type {?} */
            var uploadUrl = "/api/runtime/dfs/v1.0/formdoc";
            /** @type {?} */
            var parametersArr = originFiles.map((/**
             * @param {?} ele
             * @return {?}
             */
            function (ele) {
                return {
                    docInfo: {
                        fileContent: ele.content.substring(ele.content.indexOf(',') + 1),
                        fileName: ele.file.name,
                        index: 0,
                        total: 1,
                        metadataId: ele.id
                    },
                    formId: uploadInfo.parentDirName,
                    mode: 1,
                    rootId: uploadInfo.rootId || 'default-root'
                };
            }));
            /** @type {?} */
            var requestList = parametersArr.map((/**
             * @param {?} ele
             * @return {?}
             */
            function (ele) {
                /** @type {?} */
                var params = { body: ele };
                return apiProxy.request(mobileDevkit.HttpMethods.POST, uploadUrl, params, true);
            }));
            return rxjs.forkJoin(requestList);
        };
        /**
         * 预览附件pdf
         * @param fileType
         * @param rootId
         * @param metadataIdList
         */
        /**
         * 预览附件pdf
         * @param {?} rootId
         * @param {?} metadataIdList
         * @return {?}
         */
        UploadAttachmentService.prototype.previewPdf = /**
         * 预览附件pdf
         * @param {?} rootId
         * @param {?} metadataIdList
         * @return {?}
         */
        function (rootId, metadataIdList) {
            /** @type {?} */
            var baseUrl = '/platform/runtime/dfs/web/pdfjs/viewer.html?file=';
            /** @type {?} */
            var apiProxy = this.befRepository.apiProxy;
            /** @type {?} */
            var fileListUrl = '/api/runtime/dfs/v1.0/fileviewer/filelist';
            /** @type {?} */
            var params = {
                storageType: 2, filePath: JSON.stringify({
                    rootId: rootId,
                    metadataIdList: metadataIdList
                })
            };
            /** @type {?} */
            var body = { body: params };
            // 预览
            return apiProxy.request(mobileDevkit.HttpMethods.POST, fileListUrl, body).pipe(operators.map((/**
             * @param {?} res
             * @return {?}
             */
            function (res) { return baseUrl + encodeURIComponent(res[0].filePath); })));
        };
        /**
         * 预览office txt文档
         * @param rootId
         * @param metadataIdList
         */
        /**
         * 预览office txt文档
         * @param {?} rootId
         * @param {?} metadataIdList
         * @return {?}
         */
        UploadAttachmentService.prototype.previewDoc = /**
         * 预览office txt文档
         * @param {?} rootId
         * @param {?} metadataIdList
         * @return {?}
         */
        function (rootId, metadataIdList) {
            /** @type {?} */
            var apiProxy = this.befRepository.apiProxy;
            /** @type {?} */
            var fileListUrl = '/api/runtime/dfs/v1.0/fileviewer/filelist';
            /** @type {?} */
            var loadDocUrl = '/api/runtime/dfs/v1.0/docviewer/loaddocument';
            /** @type {?} */
            var params = {
                storageType: 2, filePath: JSON.stringify({
                    rootId: rootId,
                    metadataIdList: metadataIdList
                })
            };
            /** @type {?} */
            var body = { body: params };
            // 根据文件类型 调用不同的预览接口
            return apiProxy.request(mobileDevkit.HttpMethods.POST, fileListUrl, body)
                .pipe(operators.mergeMap((/**
             * @param {?} res
             * @return {?}
             */
            function (res) { return apiProxy.request(mobileDevkit.HttpMethods.POST, loadDocUrl, { body: { filePath: res[0].filePath } }); })), operators.map((/**
             * @param {?} res
             * @return {?}
             */
            function (res) { return res.pages; })));
        };
        /**
         * 获取图片url
         * @param attachmentId  附件id
         * @param rootId 根目录id
         * @returns
         */
        /**
         * 获取图片url
         * @param {?} attachmentId  附件id
         * @param {?} rootId 根目录id
         * @return {?}
         */
        UploadAttachmentService.prototype.getAttachmentPath = /**
         * 获取图片url
         * @param {?} attachmentId  附件id
         * @param {?} rootId 根目录id
         * @return {?}
         */
        function (attachmentId, rootId) {
            return "/api/runtime/dfs/v1.0/doc/filecontent?metadataid=" + attachmentId + "&rootid=" + rootId + "&token=" + getToken(mergeString(attachmentId, rootId));
        };
        return UploadAttachmentService;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        UploadAttachmentService.prototype.befRepository;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/lookup-mapping.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 帮助映射服务
     */
    var   /**
     * 帮助映射服务
     */
    LookupMappingService = /** @class */ (function () {
        function LookupMappingService(viewModelContext) {
            this.viewModelContext = viewModelContext;
        }
        /**
         * 映射字段
         * @summary
         * mapFields格式形如：
         * {
         *    id: "assoField.assoField",
         *    code: "assoField.assoField_Code",
         *    name: "assoField.assoField_Name"
         * }
         */
        /**
         * 映射字段
         * \@summary
         * mapFields格式形如：
         * {
         *    id: "assoField.assoField",
         *    code: "assoField.assoField_Code",
         *    name: "assoField.assoField_Name"
         * }
         * @param {?} helpData
         * @param {?} mappingFields
         * @return {?}
         */
        LookupMappingService.prototype.mapFields = /**
         * 映射字段
         * \@summary
         * mapFields格式形如：
         * {
         *    id: "assoField.assoField",
         *    code: "assoField.assoField_Code",
         *    name: "assoField.assoField_Name"
         * }
         * @param {?} helpData
         * @param {?} mappingFields
         * @return {?}
         */
        function (helpData, mappingFields) {
            var _this = this;
            console.log(helpData);
            if (!mappingFields) {
                return;
            }
            /** @type {?} */
            var helpFields = Object.keys(mappingFields);
            /** @type {?} */
            var idIndex = helpFields.findIndex((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item === 'id'; }));
            if (helpFields.includes('id') && idIndex !== 0) {
                helpFields.splice(idIndex, 1);
                helpFields = __spread(['id'], helpFields);
            }
            /** @type {?} */
            var that = this;
            helpFields.forEach((/**
             * @param {?} f
             * @return {?}
             */
            function (f) {
                // 1、获取字段值
                // 如果helpData有选中值，则获取帮助数据源里对应你字段的值；
                // 如果helpData没有值（清空场景），则返回一个空字符串
                /** @type {?} */
                var val = '';
                if (helpData) {
                    if (helpData instanceof Array) {
                        val = helpData
                            .map((/**
                         * @param {?} h
                         * @return {?}
                         */
                        function (h) {
                            return _this.getValue(f, h);
                        }))
                            .join(',');
                    }
                    else {
                        val = that.getValue(f, helpData);
                    }
                }
                // 2、设置字段值
                // 如果helpData不存在（清空场景），获取BindingData里对应字段的值，如果是数值，则设置0，其他设置上一步中的空字符串；
                // 如果helpData存在：直接设置上一步中获取的值。
                /** @type {?} */
                var pathArr = that.getBindingPathArray();
                mappingFields[f].split(',').forEach((/**
                 * @param {?} ff
                 * @return {?}
                 */
                function (ff) {
                    if (!helpData) {
                    }
                    else {
                        that.viewModelContext.bindingData.setValue(pathArr.concat(ff.split('.')), val, true, true);
                    }
                }));
            }));
        };
        /**
         * 获取字段值
         */
        /**
         * 获取字段值
         * @param {?} fieldPath
         * @return {?}
         */
        LookupMappingService.prototype.getFieldValue = /**
         * 获取字段值
         * @param {?} fieldPath
         * @return {?}
         */
        function (fieldPath) {
            /** @type {?} */
            var bindingPathArray = this.getBindingPathArray();
            /** @type {?} */
            var fieldPathArray = bindingPathArray.concat(fieldPath.split('.'));
            return this.viewModelContext.bindingData.getValue(fieldPathArray);
        };
        /**
         * @private
         * @param {?} f
         * @param {?} data
         * @return {?}
         */
        LookupMappingService.prototype.getValue = /**
         * @private
         * @param {?} f
         * @param {?} data
         * @return {?}
         */
        function (f, data) {
            /** @type {?} */
            var val = '';
            if (f.indexOf('.') === -1) {
                val = data[f];
            }
            else {
                val = f.split('.').reduce((/**
                 * @param {?} a
                 * @param {?} b
                 * @return {?}
                 */
                function (a, b) {
                    return a[b];
                }), data);
            }
            return val;
        };
        /**
         * @private
         * @return {?}
         */
        LookupMappingService.prototype.getBindingPathArray = /**
         * @private
         * @return {?}
         */
        function () {
            /** @type {?} */
            var path = this.viewModelContext.bindingData.bindingPath;
            if (path) {
                return path.split('/').filter((/**
                 * @param {?} n
                 * @return {?}
                 */
                function (n) { return n !== ''; }));
            }
            return [];
        };
        return LookupMappingService;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        LookupMappingService.prototype.viewModelContext;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/map/cloud-plus-map-operation.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var CloudPlusMapOperation = /** @class */ (function () {
        function CloudPlusMapOperation(location$) {
            this.location$ = location$;
        }
        /**
         * 云上协同  调用获取当前位置接口
         */
        /**
         * 云上协同  调用获取当前位置接口
         * @return {?}
         */
        CloudPlusMapOperation.prototype.getCurrentLoc = /**
         * 云上协同  调用获取当前位置接口
         * @return {?}
         */
        function () {
            var _this = this;
            // 具体实现
            if (window['iCityBridge']) {
                /** @type {?} */
                var iCityBridge = window['iCityBridge'];
                iCityBridge.ccworkGetLocation((/**
                 * @param {?} result
                 * @return {?}
                 */
                function (result) {
                    alert(JSON.stringify(result));
                    // 结果预处理
                    // 发布位置结果
                    _this.location$.next(_this.assembleMarkers(result));
                }));
            }
        };
        // 获取定位回调
        // 获取定位回调
        /**
         * @private
         * @param {?=} result
         * @return {?}
         */
        CloudPlusMapOperation.prototype.locationCallback = 
        // 获取定位回调
        /**
         * @private
         * @param {?=} result
         * @return {?}
         */
        function (result) {
            alert(JSON.stringify(result));
            // 结果预处理
            // 发布位置结果
            this.location$.next(this.assembleMarkers(result));
        };
        /**
         * 组装坐标结果
         * @param res 结果
         */
        /**
         * 组装坐标结果
         * @private
         * @param {?} res 结果
         * @return {?}
         */
        CloudPlusMapOperation.prototype.assembleMarkers = /**
         * 组装坐标结果
         * @private
         * @param {?} res 结果
         * @return {?}
         */
        function (res) {
            /** @type {?} */
            var loc = res.result;
            return [{ position: [Number(loc.lng), Number(loc.lat)], visible: true, draggable: true }];
        };
        return CloudPlusMapOperation;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        CloudPlusMapOperation.prototype.location$;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/map/map.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @abstract
     */
    var /**
     * @abstract
     */
    MapService = /** @class */ (function () {
        function MapService() {
            this.location$ = new rxjs.Subject();
        }
        /**
         * 获取当前位置
         */
        /**
         * 获取当前位置
         * @return {?}
         */
        MapService.prototype.getCurrentLocation = /**
         * 获取当前位置
         * @return {?}
         */
        function () {
            this.locationInstance.getCurrentLoc();
        };
        /**
         * 获取坐标集合
         */
        /**
         * 获取坐标集合
         * @return {?}
         */
        MapService.prototype.getMapMarkers = /**
         * 获取坐标集合
         * @return {?}
         */
        function () {
            return this.location$;
        };
        /**
         * @return {?}
         */
        MapService.prototype.setMapMarkers = /**
         * @return {?}
         */
        function () {
            // this.locationInstance.location$.subscribe((res) => {
            //     this.markers.next(res);
            // });
        };
        return MapService;
    }());
    if (false) {
        /** @type {?} */
        MapService.prototype.locationInstance;
        /** @type {?} */
        MapService.prototype.location$;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/map/cloud-plus-map.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var CloudPlusMapService = /** @class */ (function (_super) {
        __extends(CloudPlusMapService, _super);
        function CloudPlusMapService() {
            var _this = _super.call(this) || this;
            _this.locationInstance = new CloudPlusMapOperation(_this.location$);
            return _this;
        }
        return CloudPlusMapService;
    }(MapService));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/providers-mapper.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 映射Providers
     * @return {?}
     */
    function mapProviders() {
        return (/**
         * @return {?}
         */
        function provide() {
            /** @type {?} */
            var viewModel = this.$getViewModel();
            /** @type {?} */
            var lookupDataService = new mobileBef.BefLookupDataService(viewModel.context);
            /** @type {?} */
            var lookupMappingService = new LookupMappingService(viewModel.context);
            // 云上协同附件服务
            /** @type {?} */
            var attachmentService = new CloudPlusAttachment();
            // caf提供附件上传和预览接口调用
            /** @type {?} */
            var uploadAttachmentService = new UploadAttachmentService((/** @type {?} */ (viewModel.context.repository)));
            // 云上协同地图服务
            /** @type {?} */
            var mapService = new CloudPlusMapService();
            /** @type {?} */
            var providers = {
                LookupHttpService: lookupDataService,
                LookupMappingFn: lookupMappingService,
                uploadAttachmentService: uploadAttachmentService,
                attachmentService: attachmentService,
                mapService: mapService
            };
            return providers;
        });
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/directives/bind-listview-extender.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ɵ0$3 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        console.log(componentInstance);
        componentInstance.$bindComponentExtender(componentInstance, 'Listview');
    }, ɵ1$1 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$unbindComponentExtender(componentInstance);
    };
    /** @type {?} */
    var bindListviewExtenderDir = {
        mounted: (ɵ0$3),
        unmounted: (ɵ1$1)
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/directives/bind-lookup-extender.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ɵ0$4 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$bindComponentExtender(componentInstance, 'Lookup');
    }, ɵ1$2 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$unbindComponentExtender(componentInstance);
    };
    /** @type {?} */
    var bindLookupExtenderDir = {
        mounted: (ɵ0$4),
        unmounted: (ɵ1$2)
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/directives/bind-filter-extender.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ɵ0$5 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$bindComponentExtender(componentInstance, 'Filter');
    }, ɵ1$3 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$unbindComponentExtender(componentInstance);
    };
    /** @type {?} */
    var bindFilterExtenderDir = {
        mounted: (ɵ0$5),
        unmounted: (ɵ1$3)
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/directives/bind-tab-extender.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ɵ0$6 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$bindComponentExtender(componentInstance, 'Tab');
    }, ɵ1$4 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$unbindComponentExtender(componentInstance);
    };
    /** @type {?} */
    var bindTabExtenderDir = {
        mounted: (ɵ0$6),
        unmounted: (ɵ1$4)
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/directives/bind-search-extender.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ɵ0$7 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.children[0].children[0].componentInstance));
        componentInstance.$bindComponentExtender(componentInstance, 'Search');
    }, ɵ1$5 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.children[0].children[0].componentInstance));
        componentInstance.$unbindComponentExtender(componentInstance);
    };
    /** @type {?} */
    var bindSearchExtenderDir = {
        mounted: (ɵ0$7),
        unmounted: (ɵ1$5)
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/directives/bind-reflesh-extender.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ɵ0$8 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$bindComponentExtender(componentInstance, 'Reflesh');
    }, ɵ1$6 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$unbindComponentExtender(componentInstance);
    };
    /** @type {?} */
    var bindRefleshExtenderDir = {
        mounted: (ɵ0$8),
        unmounted: (ɵ1$6)
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/directives/bind-checkbox-extender.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ɵ0$9 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$bindComponentExtender(componentInstance, 'Checkbox');
    }, ɵ1$7 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$unbindComponentExtender(componentInstance);
    };
    /** @type {?} */
    var bindCheckboxExtenderDir = {
        mounted: (ɵ0$9),
        unmounted: (ɵ1$7)
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/directives/bind-attachment-extender.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ɵ0$a = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$bindComponentExtender(componentInstance, 'Attachment');
    }, ɵ1$8 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$unbindComponentExtender(componentInstance);
    };
    /** @type {?} */
    var bindAttachmentExtenderDir = {
        mounted: (ɵ0$a),
        unmounted: (ɵ1$8)
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/directives/map-directives/bind-map-extender.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ɵ0$b = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        /** @type {?} */
        var value = binding.value;
        componentInstance.lngPath = value.lngPath;
        componentInstance.latPath = value.latPath;
        componentInstance.$bindComponentExtender(componentInstance, 'Map');
    }, ɵ1$9 = /**
     * @param {?} el
     * @param {?} binding
     * @param {?} vnode
     * @param {?} oldVnode
     * @return {?}
     */
    function (el, binding, vnode, oldVnode) {
        /** @type {?} */
        var componentInstance = (/** @type {?} */ (vnode.componentInstance));
        componentInstance.$unbindComponentExtender(componentInstance);
    };
    /** @type {?} */
    var bindMapExtenderDir = {
        mounted: (ɵ0$b),
        // update: function (el, binding, vnode, oldVnode) {
        //     const newValue = binding.value;
        //     const componentInstance = vnode.componentInstance as any;
        //     // 监听响应式变化
        //     componentInstance.resizeEnable = newValue.resizeEnable;
        //     // 是否可以拖动
        //     componentInstance.dragEnable = newValue.dragEnable;
        //     // 初始化缩放大小
        //     componentInstance.zoom = newValue.zoom;
        //     componentInstance.zoomEnable = newValue.zoomEnable;
        //     // 地图中心点经纬度
        //     componentInstance.center = newValue.center;
        //     console.log(el);
        //     console.log(binding);
        //     console.log(vnode);
        //     console.log(oldVnode);
        // },
        unmounted: (ɵ1$9)
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/directives/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/binding-value-converters/date-converter.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 日期转换器
     */
    var   /**
     * 日期转换器
     */
    DateConverter = /** @class */ (function () {
        /**
         * 构造函数
         * @param format 日期格式
         */
        function DateConverter(format) {
            this.format = format;
        }
        /**
         * 控件值（Date对象）=>数据源（字符串）
         */
        /**
         * 控件值（Date对象）=>数据源（字符串）
         * @param {?} value
         * @return {?}
         */
        DateConverter.prototype.convertFrom = /**
         * 控件值（Date对象）=>数据源（字符串）
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (!value) {
                return '0001-01-01T00:00:00';
            }
            /** @type {?} */
            var format = this.format ? this.format : 'yyyy-MM-dd HH:mm:ss';
            return mobileDevkit.DateUtil.format(value, format);
        };
        /**
         * 数据源（字符串）=> 控件值（Date对象）
         */
        /**
         * 数据源（字符串）=> 控件值（Date对象）
         * @param {?} value
         * @return {?}
         */
        DateConverter.prototype.convertTo = /**
         * 数据源（字符串）=> 控件值（Date对象）
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (mobileDevkit.DateUtil.isEmptyDateString(value) === true) {
                return null;
            }
            return mobileDevkit.DateUtil.parse(value);
        };
        return DateConverter;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        DateConverter.prototype.format;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/binding-value-converters/array-converter.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 字符串数组转换器
     */
    var   /**
     * 字符串数组转换器
     */
    StringArrayConverter = /** @class */ (function () {
        function StringArrayConverter(separator) {
            if (separator === void 0) { separator = ','; }
            this.separator = separator;
        }
        /**
         * 控件值（字符串数组）=>数据源（字符串）
         */
        /**
         * 控件值（字符串数组）=>数据源（字符串）
         * @param {?} value
         * @return {?}
         */
        StringArrayConverter.prototype.convertFrom = /**
         * 控件值（字符串数组）=>数据源（字符串）
         * @param {?} value
         * @return {?}
         */
        function (value) {
            return value.join(this.separator);
        };
        /**
         * 数据源（字符串）=> 控件值（字符串数组）
         */
        /**
         * 数据源（字符串）=> 控件值（字符串数组）
         * @param {?} value
         * @return {?}
         */
        StringArrayConverter.prototype.convertTo = /**
         * 数据源（字符串）=> 控件值（字符串数组）
         * @param {?} value
         * @return {?}
         */
        function (value) {
            return value.split(this.separator);
        };
        return StringArrayConverter;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        StringArrayConverter.prototype.separator;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/binding-value-converters/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/attachment/attachment-filter.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var filterData = (/**
     * @param {?} rawData
     * @param {?} field
     * @param {?} rootId
     * @param {?=} isComponentAdapter
     * @return {?}
     */
    function (rawData, field, rootId, isComponentAdapter) {
        if (isComponentAdapter === void 0) { isComponentAdapter = true; }
        /** @type {?} */
        var splitFiled = field.split('/');
        /** @type {?} */
        var fields = splitFiled.filter((/**
         * @param {?} ele
         * @return {?}
         */
        function (ele) { return ele; }));
        if (!isComponentAdapter) {
            // 单条数据
            /** @type {?} */
            var data = getDataByPath(rawData, fields);
            /** @type {?} */
            var url = '/platform/common/web/assets/imgs/no-pic.png';
            // 附件不存在
            if (!data || !data.attachmentId) {
                return url;
            }
            // 附件存在
            /** @type {?} */
            var extensionName = data.fileName.substring(data.fileName.lastIndexOf('.') + 1);
            url = getAttachmentUrl(data.attachmentId, rootId, extensionName);
            return url;
        }
        /** @type {?} */
        var transformData = rawData.map((/**
         * @param {?} everyData
         * @return {?}
         */
        function (everyData) {
            /** @type {?} */
            var data = getDataByPath(everyData, fields);
            /** @type {?} */
            var url = '/platform/common/web/assets/imgs/no-pic.png';
            if (!data || !data.attachmentId) {
                return {
                    url: url,
                    id: everyData.id,
                    attachmentId: data.attachmentId,
                    name: '',
                    size: 0,
                    createTime: ''
                };
            }
            /** @type {?} */
            var extensionName = data.fileName.substring(data.fileName.lastIndexOf('.') + 1);
            url = getAttachmentUrl(data.attachmentId, rootId, extensionName);
            return {
                url: url,
                id: everyData.id,
                attachmentId: data.attachmentId,
                name: data.fileName,
                size: data.fileSize,
                createTime: data.fileCreateTime
            };
        }));
        return transformData;
    });

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/attachment/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/map/map-filter.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var filterMapData = (/**
     * @param {?} currentEntityData
     * @param {?} lng
     * @param {?} lat
     * @return {?}
     */
    function (currentEntityData, lng, lat) {
        // 经度
        /** @type {?} */
        var lngData = getDataByPath(currentEntityData, lng.split('/'));
        // 纬度
        /** @type {?} */
        var latData = getDataByPath(currentEntityData, lat.split('/'));
        return [{ position: [lngData || 117.12212, latData || 35.3243] }];
    });

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/map/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/app-creator.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var AppCreator = /** @class */ (function () {
        function AppCreator() {
        }
        /**
         * 创建Vue应用
         */
        /**
         * 创建Vue应用
         * @param {?} rootComponent
         * @return {?}
         */
        AppCreator.createApp = /**
         * 创建Vue应用
         * @param {?} rootComponent
         * @return {?}
         */
        function (rootComponent) {
            this.app = vue.createApp(rootComponent);
            // 注册内置插件
            this.app.use(DevkitPlugin);
            // 注册指令
            // this.createDirective('bind-listview-extender', bindListviewExtenderDir);
            // this.createDirective('bind-lookup-extender', bindLookupExtenderDir);
            // this.createDirective('bind-filter-extender', bindFilterExtenderDir);
            // this.createDirective('bind-serach-extender', bindSearchExtenderDir);
            // this.createDirective('bind-refresh-extender', bindRefleshExtenderDir);
            // this.createDirective('bind-tab-extender', bindTabExtenderDir);
            // this.createDirective('bind-checkbox-extender', bindCheckboxExtenderDir);
            // this.createDirective('bind-attachment-extender', bindAttachmentExtenderDir);
            // this.createDirective('bind-map-extender', bindMapExtenderDir);
            return this.app;
        };
        /**
         *
         */
        /**
         *
         * @param {?} directiveName
         * @param {?} directiveOptions
         * @return {?}
         */
        AppCreator.createDirective = /**
         *
         * @param {?} directiveName
         * @param {?} directiveOptions
         * @return {?}
         */
        function (directiveName, directiveOptions) {
            this.app.directive(directiveName, directiveOptions);
        };
        /**
         * Vue应用
         */
        AppCreator.app = null;
        return AppCreator;
    }());
    if (false) {
        /**
         * Vue应用
         * @type {?}
         */
        AppCreator.app;
    }

    exports.AppCreator = AppCreator;
    exports.BindingStates = BindingStates;
    exports.DateConverter = DateConverter;
    exports.DevkitPlugin = DevkitPlugin;
    exports.LookupMappingService = LookupMappingService;
    exports.StringArrayConverter = StringArrayConverter;
    exports.bindAttachmentExtenderDir = bindAttachmentExtenderDir;
    exports.bindCheckboxExtenderDir = bindCheckboxExtenderDir;
    exports.bindFilterExtenderDir = bindFilterExtenderDir;
    exports.bindListviewExtenderDir = bindListviewExtenderDir;
    exports.bindLookupExtenderDir = bindLookupExtenderDir;
    exports.bindMapExtenderDir = bindMapExtenderDir;
    exports.bindRefleshExtenderDir = bindRefleshExtenderDir;
    exports.bindSearchExtenderDir = bindSearchExtenderDir;
    exports.bindTabExtenderDir = bindTabExtenderDir;
    exports.filterData = filterData;
    exports.filterMapData = filterMapData;
    exports.mapBindingStates = mapBindingStates;
    exports.mapCommandMethods = mapCommandMethods;
    exports.mapCurrentEntityData = mapCurrentEntityData;
    exports.mapEntityListData = mapEntityListData;
    exports.mapForm = mapForm;
    exports.mapProviders = mapProviders;
    exports.mapUIState = mapUIState;
    exports.recomputable = recomputable;
    exports.recompute = recompute;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
