// @farris/component-querycondition/esm2015/lib/entity/controltype/checkbox/checkbox-control.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/controltype/checkbox/checkbox-control.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../../controltype";
/**
 * @record
 */
function Checkbox() { }
if (false) {
    /** @type {?} */
    Checkbox.prototype.value;
    /** @type {?} */
    Checkbox.prototype.name;
}
/**
 * 日期控件
 */
export class CheckBoxControl {
    constructor() {
        /* radio 数组 */
        this.data = [];
        /* 分隔符 默认逗号*/
        this.separator = ',';
        /* 值类型是否是string */
        this.isStringValue = true;
    }
    /**
     * 获取控件类型
     * @return {?}
     */
    getControlType() {
        return ControlType.BoolCheck;
    }
    /**
     * @return {?}
     */
    isRequired() {
        return !!this.require;
    }
}
if (false) {
    /**
     * 自定义样式
     * @type {?}
     */
    CheckBoxControl.prototype.className;
    /** @type {?} */
    CheckBoxControl.prototype.data;
    /** @type {?} */
    CheckBoxControl.prototype.name;
    /** @type {?} */
    CheckBoxControl.prototype.horizontal;
    /** @type {?} */
    CheckBoxControl.prototype.disable;
    /** @type {?} */
    CheckBoxControl.prototype.value;
    /** @type {?} */
    CheckBoxControl.prototype.separator;
    /** @type {?} */
    CheckBoxControl.prototype.isStringValue;
    /** @type {?} */
    CheckBoxControl.prototype.require;
}
