// @gsp-cmp/webcommand/esm5/lib/vm/BranchCommandItem.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommandItemType } from "./CommandItemType";
import { CommandItemConvertor } from "./ICommandItem";
var BranchCommandItem = /** @class */ (function () {
    function BranchCommandItem() {
    }
    /**
     * @return {?}
     */
    BranchCommandItem.prototype.GetItemId = /**
     * @return {?}
     */
    function () {
        return this.Id;
    };
    /**
     * @return {?}
     */
    BranchCommandItem.prototype.GetItemCode = /**
     * @return {?}
     */
    function () {
        return this.Code;
    };
    /**
     * @return {?}
     */
    BranchCommandItem.prototype.GetItemName = /**
     * @return {?}
     */
    function () {
        return this.Name;
    };
    /**
     * @return {?}
     */
    BranchCommandItem.prototype.GetItemType = /**
     * @return {?}
     */
    function () {
        return CommandItemType.Branch;
    };
    return BranchCommandItem;
}());
export { BranchCommandItem };
if (false) {
    /** @type {?} */
    BranchCommandItem.prototype.Id;
    /** @type {?} */
    BranchCommandItem.prototype.Code;
    /** @type {?} */
    BranchCommandItem.prototype.Name;
    /** @type {?} */
    BranchCommandItem.prototype.ConditionType;
    /** @type {?} */
    BranchCommandItem.prototype.Express;
    /** @type {?} */
    BranchCommandItem.prototype.Items;
}
var BranchCommandItemConvertor = /** @class */ (function () {
    function BranchCommandItemConvertor() {
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    BranchCommandItemConvertor.prototype.ConvertJObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var branchItem = (/** @type {?} */ (obj));
        /** @type {?} */
        var jobj = new Object();
        jobj["Id"] = branchItem.Id;
        jobj["Code"] = branchItem.Code;
        jobj["Name"] = branchItem.Name;
        jobj["ConditionType"] = branchItem.ConditionType;
        jobj["Express"] = branchItem.Express;
        if (branchItem.Items != null) {
            /** @type {?} */
            var itemArray_1 = [];
            /** @type {?} */
            var itemConvertor_1 = new CommandItemConvertor();
            branchItem.Items.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                itemArray_1.push(itemConvertor_1.ConvertJObject(element));
            }));
            jobj["Items"] = itemArray_1;
        }
        return jobj;
    };
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    BranchCommandItemConvertor.prototype.InitFromJobject = /**
     * @param {?} jsonObj
     * @return {?}
     */
    function (jsonObj) {
        /** @type {?} */
        var branchItem = new BranchCommandItem();
        branchItem.Id = jsonObj["Id"];
        branchItem.Code = jsonObj["Code"];
        branchItem.Name = jsonObj["Name"];
        branchItem.ConditionType = jsonObj["ConditionType"];
        branchItem.Express = jsonObj["Express"];
        if (jsonObj["Items"] != null) {
            branchItem.Items = new Array();
            /** @type {?} */
            var itemConvertor_2 = new CommandItemConvertor();
            jsonObj["Items"].forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                branchItem.Items.push(itemConvertor_2.InitFromJobject(element));
            }));
        }
        return branchItem;
    };
    return BranchCommandItemConvertor;
}());
export { BranchCommandItemConvertor };
