// @gsp-cmp/webcommand/esm5/lib/vm/BranchCollectionCommandItem.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommandItemType } from "./CommandItemType";
import { BranchCommandItemConvertor } from "./BranchCommandItem";
var BranchCollectionCommandItem = /** @class */ (function () {
    function BranchCollectionCommandItem() {
    }
    /**
     * @return {?}
     */
    BranchCollectionCommandItem.prototype.GetItemId = /**
     * @return {?}
     */
    function () {
        return this.Id;
    };
    /**
     * @return {?}
     */
    BranchCollectionCommandItem.prototype.GetItemCode = /**
     * @return {?}
     */
    function () {
        return this.Code;
    };
    /**
     * @return {?}
     */
    BranchCollectionCommandItem.prototype.GetItemName = /**
     * @return {?}
     */
    function () {
        return this.Name;
    };
    /**
     * @return {?}
     */
    BranchCollectionCommandItem.prototype.GetItemType = /**
     * @return {?}
     */
    function () {
        return CommandItemType.BranchCollection;
    };
    return BranchCollectionCommandItem;
}());
export { BranchCollectionCommandItem };
if (false) {
    /** @type {?} */
    BranchCollectionCommandItem.prototype.Id;
    /** @type {?} */
    BranchCollectionCommandItem.prototype.Code;
    /** @type {?} */
    BranchCollectionCommandItem.prototype.Name;
    /** @type {?} */
    BranchCollectionCommandItem.prototype.Items;
}
var BranchCollectionCommandItemConvertor = /** @class */ (function () {
    function BranchCollectionCommandItemConvertor() {
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    BranchCollectionCommandItemConvertor.prototype.ConvertJObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var branchCollection = (/** @type {?} */ (obj));
        /** @type {?} */
        var jobj = new Object();
        jobj["Id"] = branchCollection.Id;
        jobj["Code"] = branchCollection.Code;
        jobj["Name"] = branchCollection.Name;
        if (branchCollection.Items != null) {
            /** @type {?} */
            var itemArray_1 = [];
            /** @type {?} */
            var itemConvertor_1 = new BranchCommandItemConvertor();
            branchCollection.Items.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                itemArray_1.push(itemConvertor_1.ConvertJObject(element));
            }));
            jobj["Items"] = itemArray_1;
        }
        return jobj;
    };
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    BranchCollectionCommandItemConvertor.prototype.InitFromJobject = /**
     * @param {?} jsonObj
     * @return {?}
     */
    function (jsonObj) {
        /** @type {?} */
        var branchCollectionItem = new BranchCollectionCommandItem();
        branchCollectionItem.Id = jsonObj["Id"];
        branchCollectionItem.Code = jsonObj["Code"];
        branchCollectionItem.Name = jsonObj["Name"];
        if (jsonObj["Items"] != null) {
            branchCollectionItem.Items = new Array();
            /** @type {?} */
            var itemConvertor_2 = new BranchCommandItemConvertor();
            jsonObj["Items"].forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                branchCollectionItem.Items.push(itemConvertor_2.InitFromJobject(element));
            }));
        }
        return branchCollectionItem;
    };
    return BranchCollectionCommandItemConvertor;
}());
export { BranchCollectionCommandItemConvertor };
