// @gsp-cmp/webcommand/esm5/lib/vm/CmpMethodParamConfig.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CmpMethodParamConfig = /** @class */ (function () {
    function CmpMethodParamConfig() {
    }
    return CmpMethodParamConfig;
}());
export { CmpMethodParamConfig };
if (false) {
    /** @type {?} */
    CmpMethodParamConfig.prototype.ParamCode;
    /** @type {?} */
    CmpMethodParamConfig.prototype.ParamName;
    /** @type {?} */
    CmpMethodParamConfig.prototype.ParamExpress;
}
var CmpMethodParamConfigConvert = /** @class */ (function () {
    function CmpMethodParamConfigConvert() {
    }
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    CmpMethodParamConfigConvert.prototype.InitFromJobject = /**
     * @param {?} jsonObj
     * @return {?}
     */
    function (jsonObj) {
        /** @type {?} */
        var cmpMethodParamConfig = new CmpMethodParamConfig();
        cmpMethodParamConfig.ParamCode = jsonObj["ParamCode"];
        cmpMethodParamConfig.ParamName = jsonObj["ParamName"];
        cmpMethodParamConfig.ParamExpress = jsonObj["ParamExpress"];
        return cmpMethodParamConfig;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CmpMethodParamConfigConvert.prototype.ConvertJObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var cmpMethodParamConfig = (/** @type {?} */ (obj));
        /** @type {?} */
        var jobj = new Object();
        jobj["ParamCode"] = cmpMethodParamConfig.ParamCode;
        jobj["ParamName"] = cmpMethodParamConfig.ParamName;
        jobj["ParamExpress"] = cmpMethodParamConfig.ParamExpress;
        return jobj;
    };
    return CmpMethodParamConfigConvert;
}());
export { CmpMethodParamConfigConvert };
