// @gsp-cmp/webcommand/esm5/lib/vm/CmpMethodRefering.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CmpMethodParamConfigConvert } from "./CmpMethodParamConfig";
import { CommandItemType } from "./CommandItemType";
var CmpMethodRefering = /** @class */ (function () {
    function CmpMethodRefering() {
    }
    /**
     * @return {?}
     */
    CmpMethodRefering.prototype.GetItemType = /**
     * @return {?}
     */
    function () {
        return CommandItemType.MethodRefer;
    };
    /**
     * @return {?}
     */
    CmpMethodRefering.prototype.GetItemCode = /**
     * @return {?}
     */
    function () {
        return this.Code;
    };
    /**
     * @return {?}
     */
    CmpMethodRefering.prototype.GetItemName = /**
     * @return {?}
     */
    function () {
        return this.Name;
    };
    /**
     * @return {?}
     */
    CmpMethodRefering.prototype.GetItemId = /**
     * @return {?}
     */
    function () {
        return this.Id;
    };
    return CmpMethodRefering;
}());
export { CmpMethodRefering };
if (false) {
    /** @type {?} */
    CmpMethodRefering.prototype.Id;
    /** @type {?} */
    CmpMethodRefering.prototype.Code;
    /** @type {?} */
    CmpMethodRefering.prototype.Name;
    /** @type {?} */
    CmpMethodRefering.prototype.ComponentId;
    /** @type {?} */
    CmpMethodRefering.prototype.ComponentCode;
    /** @type {?} */
    CmpMethodRefering.prototype.ComponentName;
    /** @type {?} */
    CmpMethodRefering.prototype.ComponentPath;
    /** @type {?} */
    CmpMethodRefering.prototype.MethodId;
    /** @type {?} */
    CmpMethodRefering.prototype.MethodCode;
    /** @type {?} */
    CmpMethodRefering.prototype.MethodName;
    /** @type {?} */
    CmpMethodRefering.prototype.IsReplaced;
    /** @type {?} */
    CmpMethodRefering.prototype.IsBeforeExpansion;
    /** @type {?} */
    CmpMethodRefering.prototype.IsAfterExpansion;
    /** @type {?} */
    CmpMethodRefering.prototype.ParamConfigs;
}
var CmpMethodReferingConvert = /** @class */ (function () {
    function CmpMethodReferingConvert() {
    }
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    CmpMethodReferingConvert.prototype.InitFromJobject = /**
     * @param {?} jsonObj
     * @return {?}
     */
    function (jsonObj) {
        /** @type {?} */
        var cmpMethodRefering = new CmpMethodRefering();
        cmpMethodRefering.Id = jsonObj["Id"];
        cmpMethodRefering.Code = jsonObj["Code"];
        cmpMethodRefering.Name = jsonObj["Name"];
        cmpMethodRefering.MethodId = jsonObj["MethodId"];
        cmpMethodRefering.MethodCode = jsonObj["MethodCode"];
        cmpMethodRefering.MethodName = jsonObj["MethodName"];
        cmpMethodRefering.ComponentId = jsonObj["ComponentId"];
        cmpMethodRefering.ComponentCode = jsonObj["ComponentCode"];
        cmpMethodRefering.ComponentName = jsonObj["ComponentName"];
        cmpMethodRefering.ComponentPath = jsonObj["ComponentPath"];
        cmpMethodRefering.IsReplaced = jsonObj["IsReplaced"];
        cmpMethodRefering.IsBeforeExpansion = jsonObj["IsBeforeExpansion"];
        cmpMethodRefering.IsAfterExpansion = jsonObj["IsAfterExpansion"];
        if (jsonObj["ParamConfigs"] != null) {
            cmpMethodRefering.ParamConfigs = new Array();
            /** @type {?} */
            var convertor_1 = new CmpMethodParamConfigConvert();
            jsonObj["ParamConfigs"].forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                cmpMethodRefering.ParamConfigs.push(convertor_1.InitFromJobject(element));
            }));
        }
        return cmpMethodRefering;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CmpMethodReferingConvert.prototype.ConvertJObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var cmpMethodRefering = (/** @type {?} */ (obj));
        /** @type {?} */
        var jobj = new Object();
        jobj["Id"] = cmpMethodRefering.Id;
        jobj["Code"] = cmpMethodRefering.Code;
        jobj["Name"] = cmpMethodRefering.Name;
        jobj["MethodId"] = cmpMethodRefering.MethodId;
        jobj["MethodCode"] = cmpMethodRefering.MethodCode;
        jobj["MethodName"] = cmpMethodRefering.MethodName;
        jobj["ComponentId"] = cmpMethodRefering.ComponentId;
        jobj["ComponentCode"] = cmpMethodRefering.ComponentCode;
        jobj["ComponentName"] = cmpMethodRefering.ComponentName;
        jobj["ComponentPath"] = cmpMethodRefering.ComponentPath;
        jobj["IsReplaced"] = cmpMethodRefering.IsReplaced;
        jobj["IsBeforeExpansion"] = cmpMethodRefering.IsBeforeExpansion;
        jobj["IsAfterExpansion"] = cmpMethodRefering.IsAfterExpansion;
        if (cmpMethodRefering.ParamConfigs != null) {
            /** @type {?} */
            var CommandsJArry_1 = jobj["ParamConfigs"] = [];
            /** @type {?} */
            var convertor_2 = new CmpMethodParamConfigConvert();
            cmpMethodRefering.ParamConfigs.forEach((/**
             * @param {?} paramConfig
             * @return {?}
             */
            function (paramConfig) {
                CommandsJArry_1.push(convertor_2.ConvertJObject(paramConfig));
            }));
        }
        return jobj;
    };
    return CmpMethodReferingConvert;
}());
export { CmpMethodReferingConvert };
