// @gsp-cmp/webcommand/esm5/lib/vm/Command.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommandItemConvertor } from "./ICommandItem";
var WebCommand = /** @class */ (function () {
    function WebCommand() {
    }
    return WebCommand;
}());
export { WebCommand };
if (false) {
    /** @type {?} */
    WebCommand.prototype.Id;
    /** @type {?} */
    WebCommand.prototype.Code;
    /** @type {?} */
    WebCommand.prototype.Name;
    /** @type {?} */
    WebCommand.prototype.Description;
    /** @type {?} */
    WebCommand.prototype.Parameters;
    /** @type {?} */
    WebCommand.prototype.SourceCode;
    /** @type {?} */
    WebCommand.prototype.Items;
}
var WebCommandConvertor = /** @class */ (function () {
    function WebCommandConvertor() {
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    WebCommandConvertor.prototype.ConvertJObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var cmpOp = (/** @type {?} */ (obj));
        /** @type {?} */
        var jobj = new Object();
        jobj["Id"] = cmpOp.Id;
        jobj["Code"] = cmpOp.Code;
        jobj["Name"] = cmpOp.Name;
        jobj["Description"] = cmpOp.Description;
        jobj["SourceCode"] = cmpOp.SourceCode;
        jobj["Parameters"] = cmpOp.Parameters;
        /** @type {?} */
        var items = [];
        if (cmpOp.Items != null) {
            /** @type {?} */
            var itemConvertor_1 = new CommandItemConvertor();
            cmpOp.Items.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                items.push(itemConvertor_1.ConvertJObject(element));
            }));
        }
        jobj["Items"] = items;
        return jobj;
    };
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    WebCommandConvertor.prototype.InitFromJobject = /**
     * @param {?} jsonObj
     * @return {?}
     */
    function (jsonObj) {
        /** @type {?} */
        var cmpOp = new WebCommand();
        cmpOp.Id = jsonObj["Id"];
        cmpOp.Code = jsonObj["Code"];
        cmpOp.Name = jsonObj["Name"];
        cmpOp.Description = jsonObj["Description"];
        cmpOp.SourceCode = jsonObj["SourceCode"];
        cmpOp.Parameters = jsonObj["Parameters"];
        cmpOp.Items = new Array();
        if (jsonObj["Items"] != null) {
            /** @type {?} */
            var itemConvertor_2 = new CommandItemConvertor();
            jsonObj["Items"].forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                cmpOp.Items.push(itemConvertor_2.InitFromJobject(element));
            }));
        }
        return cmpOp;
    };
    return WebCommandConvertor;
}());
export { WebCommandConvertor };
