// @gsp-cmp/webcommand/esm5/lib/vm/ExtendProperty.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ExtendProperty = /** @class */ (function () {
    function ExtendProperty() {
    }
    return ExtendProperty;
}());
export { ExtendProperty };
if (false) {
    /** @type {?} */
    ExtendProperty.prototype.FormCode;
    /** @type {?} */
    ExtendProperty.prototype.IsCommon;
}
var ExtendsConvert = /** @class */ (function () {
    function ExtendsConvert() {
    }
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    ExtendsConvert.prototype.InitFromJobject = /**
     * @param {?} jsonObj
     * @return {?}
     */
    function (jsonObj) {
        if (jsonObj) {
            /** @type {?} */
            var extendProperty = new ExtendProperty();
            extendProperty.FormCode = jsonObj["FormCode"];
            extendProperty.IsCommon = jsonObj["IsCommon"];
            return extendProperty;
        }
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ExtendsConvert.prototype.ConvertJObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var extendProperty = (/** @type {?} */ (obj));
        /** @type {?} */
        var jobj = new Object();
        jobj["FormCode"] = extendProperty.FormCode;
        jobj["IsCommon"] = extendProperty.IsCommon;
        return jobj;
    };
    return ExtendsConvert;
}());
export { ExtendsConvert };
