// @gsp-cmp/webcommand/esm5/lib/vm/ICommandItem.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommandItemType } from "./CommandItemType";
import { CmpMethodRefering } from "./CmpMethodRefering";
import { BranchCommandItemConvertor } from "./BranchCommandItem";
import { BranchCollectionCommandItemConvertor } from "./BranchCollectionCommandItem";
/**
 * @record
 */
export function ICommandItem() { }
if (false) {
    /**
     * @return {?}
     */
    ICommandItem.prototype.GetItemType = function () { };
    /**
     * @return {?}
     */
    ICommandItem.prototype.GetItemCode = function () { };
    /**
     * @return {?}
     */
    ICommandItem.prototype.GetItemName = function () { };
    /**
     * @return {?}
     */
    ICommandItem.prototype.GetItemId = function () { };
}
var CommandItemConvertor = /** @class */ (function () {
    function CommandItemConvertor() {
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    CommandItemConvertor.prototype.ConvertJObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var commandItem = (/** @type {?} */ (obj));
        /** @type {?} */
        var itemType = commandItem.GetItemType();
        /** @type {?} */
        var jobj = new Object();
        jobj["Type"] = itemType;
        if (itemType == CommandItemType.MethodRefer) {
            jobj["Content"] = commandItem;
        }
        else if (itemType == CommandItemType.Branch) {
            /** @type {?} */
            var branchConvertor = new BranchCommandItemConvertor();
            jobj["Content"] = branchConvertor.ConvertJObject((/** @type {?} */ (commandItem)));
        }
        else if (itemType == CommandItemType.BranchCollection) {
            /** @type {?} */
            var branchCollectionConvertor = new BranchCollectionCommandItemConvertor();
            jobj["Content"] = branchCollectionConvertor.ConvertJObject((/** @type {?} */ (commandItem)));
        }
        return jobj;
    };
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    CommandItemConvertor.prototype.InitFromJobject = /**
     * @param {?} jsonObj
     * @return {?}
     */
    function (jsonObj) {
        /** @type {?} */
        var itemType = (/** @type {?} */ (jsonObj["Type"]));
        /** @type {?} */
        var content = jsonObj["Content"];
        if (itemType == CommandItemType.MethodRefer) {
            return Object.assign(new CmpMethodRefering(), (/** @type {?} */ (content)));
            ;
        }
        else if (itemType == CommandItemType.Branch) {
            /** @type {?} */
            var branchConvertor = new BranchCommandItemConvertor();
            return branchConvertor.InitFromJobject(jsonObj["Content"]);
        }
        else if (itemType == CommandItemType.BranchCollection) {
            /** @type {?} */
            var branchCollectionConvertor = new BranchCollectionCommandItemConvertor();
            return branchCollectionConvertor.InitFromJobject(jsonObj["Content"]);
        }
        return null;
    };
    return CommandItemConvertor;
}());
export { CommandItemConvertor };
