// @gsp-cmp/webcommand/esm5/lib/vm/WebCommandMetadata.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { WebCommandConvertor } from "./Command";
import { ExtendsConvert } from "./ExtendProperty";
var WebCommandMetadata = /** @class */ (function () {
    function WebCommandMetadata() {
    }
    return WebCommandMetadata;
}());
export { WebCommandMetadata };
if (false) {
    /** @type {?} */
    WebCommandMetadata.prototype.Id;
    /** @type {?} */
    WebCommandMetadata.prototype.Code;
    /** @type {?} */
    WebCommandMetadata.prototype.Name;
    /** @type {?} */
    WebCommandMetadata.prototype.Description;
    /** @type {?} */
    WebCommandMetadata.prototype.Extends;
    /** @type {?} */
    WebCommandMetadata.prototype.Commands;
}
var WebCommandMetadataConvertor = /** @class */ (function () {
    function WebCommandMetadataConvertor() {
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    WebCommandMetadataConvertor.prototype.ConvertJObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var metadata = (/** @type {?} */ (obj));
        /** @type {?} */
        var jobj = new Object();
        jobj["Id"] = metadata.Id;
        jobj["Code"] = metadata.Code;
        jobj["Name"] = metadata.Name;
        jobj["Description"] = metadata.Description;
        if (metadata.Commands != null) {
            /** @type {?} */
            var CommandsJArry_1 = jobj["Commands"] = [];
            /** @type {?} */
            var convertor_1 = new WebCommandConvertor();
            metadata.Commands.forEach((/**
             * @param {?} command
             * @return {?}
             */
            function (command) {
                CommandsJArry_1.push(convertor_1.ConvertJObject(command));
            }));
        }
        if (metadata.Extends != null) {
            /** @type {?} */
            var convertor = new ExtendsConvert();
            jobj["Extends"] = convertor.ConvertJObject(metadata.Extends);
        }
        return jobj;
    };
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    WebCommandMetadataConvertor.prototype.InitFromJobject = /**
     * @param {?} jsonObj
     * @return {?}
     */
    function (jsonObj) {
        /** @type {?} */
        var metadata = new WebCommandMetadata();
        metadata.Id = jsonObj["Id"];
        metadata.Code = jsonObj["Code"];
        metadata.Name = jsonObj["Name"];
        metadata.Description = jsonObj["Description"];
        /** @type {?} */
        var CommandsJArry = jsonObj["Commands"];
        if (CommandsJArry != null) {
            metadata.Commands = new Array();
            CommandsJArry.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                /** @type {?} */
                var cmpOpSerializer = new WebCommandConvertor();
                metadata.Commands.push((/** @type {?} */ (cmpOpSerializer.InitFromJobject(element))));
            }));
        }
        /** @type {?} */
        var convertor = new ExtendsConvert();
        metadata.Extends = convertor.InitFromJobject(jsonObj["Extends"]);
        return metadata;
    };
    return WebCommandMetadataConvertor;
}());
export { WebCommandMetadataConvertor };
