// @gsp-lcm/bo-tree/esm5/lib/bo-tree.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bo-tree.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TreeNodeContentDirective } from './directives/tree-node-content.directive';
import { BOTreePanelComponent } from './views/bo-tree-panel/bo-tree-panel.component';
import { BOTreePanelState } from './views/bo-tree-panel/bo-tree-panel.state';
import { HttpService } from '@ecp-caf/caf-common';
import { BusinessObjectRtService } from "@gsp-lcm/bo-rt-service";
import { TreeTableModule } from '@farris/ui-treetable';
var BOTreePanelPackageModule = /** @class */ (function () {
    function BOTreePanelPackageModule() {
    }
    BOTreePanelPackageModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        TreeNodeContentDirective,
                        BOTreePanelComponent
                    ],
                    imports: [
                        CommonModule,
                        TreeTableModule
                    ],
                    exports: [BOTreePanelComponent],
                    providers: [BOTreePanelState, HttpService, BusinessObjectRtService],
                    entryComponents: [BOTreePanelComponent]
                },] }
    ];
    return BOTreePanelPackageModule;
}());
export { BOTreePanelPackageModule };
