// @farris/mobile-command-services/esm5/lib/validator-services/validator.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/validator-services/validator.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { EMPTY, of, zip } from 'rxjs';
import { switchMap } from 'rxjs/operators';
/**
 * 数据服务基类
 */
var /**
 * 数据服务基类
 */
ValidatorServices = /** @class */ (function () {
    function ValidatorServices(viewModelContext) {
        this.viewModelContext = viewModelContext;
    }
    /**
     * 获取某个视图模型上得所有验证结果
     * @param viewModelId 视图模型id
     */
    /**
     * 获取某个视图模型上得所有验证结果
     * @param {?=} viewModelId 视图模型id
     * @return {?}
     */
    ValidatorServices.prototype.validateFields = /**
     * 获取某个视图模型上得所有验证结果
     * @param {?=} viewModelId 视图模型id
     * @return {?}
     */
    function (viewModelId) {
        if (viewModelId) {
            return this.viewModelContext.appContext.viewModelContextManager.getContextById(viewModelId).form.validateFields();
        }
        else {
            return this.viewModelContext.form.validateFields();
        }
    };
    /**
     * 处理某个视图模型上得所有验证结果
     * @param viewModelId 视图模型id
     */
    /**
     * 处理某个视图模型上得所有验证结果
     * @param {?=} viewModelId 视图模型id
     * @return {?}
     */
    ValidatorServices.prototype.handleValidateFields = /**
     * 处理某个视图模型上得所有验证结果
     * @param {?=} viewModelId 视图模型id
     * @return {?}
     */
    function (viewModelId) {
        /** @type {?} */
        var validateFieldsResults = [];
        if (viewModelId) {
            validateFieldsResults = this.viewModelContext.appContext.viewModelContextManager.getContextById(viewModelId).form.validateFields();
        }
        else {
            this.viewModelContext.appContext.viewModelContextManager.getRootContextAndPosterityById(this.viewModelContext.id).forEach((/**
             * @param {?} context
             * @return {?}
             */
            function (context) {
                validateFieldsResults = validateFieldsResults.concat(context.form.validateFields());
            }));
        }
        if (validateFieldsResults.length === 0) {
            return of(true);
        }
        return zip.apply(void 0, tslib_1.__spread(validateFieldsResults)).pipe(switchMap((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            /** @type {?} */
            var index = value.findIndex((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item['passing'] === false; }));
            if (index > -1) {
                return EMPTY;
            }
            else {
                return of(true);
            }
        })));
    };
    /**
     * 获取某个字段的验证结果
     * @param name 字段名称
     * @param viewModelId 视图模型id
     */
    /**
     * 获取某个字段的验证结果
     * @param {?} name 字段名称
     * @param {?=} viewModelId 视图模型id
     * @return {?}
     */
    ValidatorServices.prototype.getFieldError = /**
     * 获取某个字段的验证结果
     * @param {?} name 字段名称
     * @param {?=} viewModelId 视图模型id
     * @return {?}
     */
    function (name, viewModelId) {
        if (viewModelId) {
            return this.viewModelContext.appContext.viewModelContextManager.getContextById(viewModelId).form.getFieldError(name);
        }
        else {
            return this.viewModelContext.form.getFieldError(name);
        }
    };
    /**
     * 重置一组校验状态 不传参数清除所有
     * @param fields name的数组
     * @param viewModelId 视图模型id
     */
    /**
     * 重置一组校验状态 不传参数清除所有
     * @param {?=} fields name的数组
     * @param {?=} viewModelId 视图模型id
     * @return {?}
     */
    ValidatorServices.prototype.resetFieldsValidate = /**
     * 重置一组校验状态 不传参数清除所有
     * @param {?=} fields name的数组
     * @param {?=} viewModelId 视图模型id
     * @return {?}
     */
    function (fields, viewModelId) {
        if (fields && typeof fields === 'string') {
            fields = JSON.parse(fields);
        }
        if (viewModelId) {
            return this.viewModelContext.appContext.viewModelContextManager.getContextById(viewModelId).form.resetFieldsValidate(fields);
        }
        else {
            this.viewModelContext.appContext.viewModelContextManager.getRootContextAndPosterityById(this.viewModelContext.id).forEach((/**
             * @param {?} context
             * @return {?}
             */
            function (context) {
                context.form.resetFieldsValidate(fields);
            }));
        }
    };
    return ValidatorServices;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    ValidatorServices.prototype.viewModelContext;
}
export { ValidatorServices };
