// @gsp-svc/expression/esm5/lib/expression-context-entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 内部上下文实体
 * \@property sessionEntities 预置的系统变量列表
 * \@property contextEntities 预置的上下文变量列表
 * \@property funcEntities 预置的函数列表
 * \@property expr 预置的表达式文本
 * \@property schemaMap 预置schema键值对
 * \@property newSchemaMap 预置newSchema键值对
 * \@property injectedEntities 注入实体对象
 * \@property smartHelpEntities 注入帮助
 */
var /**
 * 内部上下文实体
 * \@property sessionEntities 预置的系统变量列表
 * \@property contextEntities 预置的上下文变量列表
 * \@property funcEntities 预置的函数列表
 * \@property expr 预置的表达式文本
 * \@property schemaMap 预置schema键值对
 * \@property newSchemaMap 预置newSchema键值对
 * \@property injectedEntities 注入实体对象
 * \@property smartHelpEntities 注入帮助
 */
ExpressionContextEntity = /** @class */ (function () {
    function ExpressionContextEntity() {
        this.sessionEntities = [];
        this.contextEntities = [];
        this.funcEntities = [];
        this.expr = null;
        this.sexpr = "";
        this.schemaMap = new Map();
        this.newSchemaMap = new Map();
        this.injectedEntities = [];
        this.smartHelpEntities = [];
    }
    return ExpressionContextEntity;
}());
/**
 * 内部上下文实体
 * \@property sessionEntities 预置的系统变量列表
 * \@property contextEntities 预置的上下文变量列表
 * \@property funcEntities 预置的函数列表
 * \@property expr 预置的表达式文本
 * \@property schemaMap 预置schema键值对
 * \@property newSchemaMap 预置newSchema键值对
 * \@property injectedEntities 注入实体对象
 * \@property smartHelpEntities 注入帮助
 */
export { ExpressionContextEntity };
if (false) {
    /** @type {?} */
    ExpressionContextEntity.prototype.sessionEntities;
    /** @type {?} */
    ExpressionContextEntity.prototype.contextEntities;
    /** @type {?} */
    ExpressionContextEntity.prototype.funcEntities;
    /** @type {?} */
    ExpressionContextEntity.prototype.expr;
    /** @type {?} */
    ExpressionContextEntity.prototype.sexpr;
    /** @type {?} */
    ExpressionContextEntity.prototype.schemaMap;
    /** @type {?} */
    ExpressionContextEntity.prototype.newSchemaMap;
    /** @type {?} */
    ExpressionContextEntity.prototype.injectedEntities;
    /** @type {?} */
    ExpressionContextEntity.prototype.smartHelpEntities;
}
