// @gsp-svc/expression/esm5/lib/expression-event-service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * 自定义的表达式事件服务类
 * 功能：定义表达式内部的事件发布与监听接口
 */
var ExpressionEventService = /** @class */ (function () {
    function ExpressionEventService() {
        // Observable string sources
        this.windowClosed = new Subject();
        // Observable string streams
        this.windowClosed$ = this.windowClosed.asObservable();
        //创建Observable源
        this._expressionsource = new Subject();
        //在光标处插入表达式事件发布与订阅
        this._insertexpression = new Subject();
        this._clearexpression = new Subject();
        //函数选择焦点清除
        this._clearFuncSelect = new Subject();
        //变量选择焦点清除
        this._clearVarSelect = new Subject();
    }
    // Service message commands
    // Service message commands
    /**
     * @return {?}
     */
    ExpressionEventService.prototype.closeWindow = 
    // Service message commands
    /**
     * @return {?}
     */
    function () {
        this.windowClosed.next();
    };
    //定义发布表达式实体方法
    //定义发布表达式实体方法
    /**
     * @param {?} expressionentity
     * @return {?}
     */
    ExpressionEventService.prototype.setExpression = 
    //定义发布表达式实体方法
    /**
     * @param {?} expressionentity
     * @return {?}
     */
    function (expressionentity) {
        this._expressionsource.next(expressionentity);
    };
    //定义获得表达式实体方法
    //定义获得表达式实体方法
    /**
     * @return {?}
     */
    ExpressionEventService.prototype.getExpression = 
    //定义获得表达式实体方法
    /**
     * @return {?}
     */
    function () {
        return this._expressionsource.asObservable();
    };
    /**
     * @param {?} expression
     * @return {?}
     */
    ExpressionEventService.prototype.insertExpression = /**
     * @param {?} expression
     * @return {?}
     */
    function (expression) {
        this._insertexpression.next(expression);
    };
    /**
     * @return {?}
     */
    ExpressionEventService.prototype.currentExpression = /**
     * @return {?}
     */
    function () {
        return this._insertexpression.asObservable();
    };
    /**
     * @return {?}
     */
    ExpressionEventService.prototype.pushclearExpression = /**
     * @return {?}
     */
    function () {
        this._clearexpression.next();
    };
    /**
     * @return {?}
     */
    ExpressionEventService.prototype.clearExpression = /**
     * @return {?}
     */
    function () {
        return this._clearexpression.asObservable();
    };
    /**
     * @return {?}
     */
    ExpressionEventService.prototype.clearFuncSelect = /**
     * @return {?}
     */
    function () {
        this._clearFuncSelect.next();
    };
    /**
     * @return {?}
     */
    ExpressionEventService.prototype.subClearFuncSelect = /**
     * @return {?}
     */
    function () {
        return this._clearFuncSelect.asObservable();
    };
    /**
     * @return {?}
     */
    ExpressionEventService.prototype.clearVarSelect = /**
     * @return {?}
     */
    function () {
        this._clearFuncSelect.next();
    };
    /**
     * @return {?}
     */
    ExpressionEventService.prototype.subClearVarSelect = /**
     * @return {?}
     */
    function () {
        return this._clearFuncSelect.asObservable();
    };
    ExpressionEventService.decorators = [
        { type: Injectable }
    ];
    return ExpressionEventService;
}());
export { ExpressionEventService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ExpressionEventService.prototype.windowClosed;
    /** @type {?} */
    ExpressionEventService.prototype.windowClosed$;
    /**
     * @type {?}
     * @private
     */
    ExpressionEventService.prototype._expressionsource;
    /**
     * @type {?}
     * @private
     */
    ExpressionEventService.prototype._insertexpression;
    /**
     * @type {?}
     * @private
     */
    ExpressionEventService.prototype._clearexpression;
    /**
     * @type {?}
     * @private
     */
    ExpressionEventService.prototype._clearFuncSelect;
    /**
     * @type {?}
     * @private
     */
    ExpressionEventService.prototype._clearVarSelect;
}
