// @gsp-svc/expression/esm5/lib/funcentity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 函数实体
 * \@property key 函数语义化名称
 * \@property params 函数参数集
 * \@property name 函数名称
 * \@property type 函数类型
 * \@property className 所在类名
 * \@property description 节点自描述
 * @author linchang 2018/12/7
 */
var /**
 * 函数实体
 * \@property key 函数语义化名称
 * \@property params 函数参数集
 * \@property name 函数名称
 * \@property type 函数类型
 * \@property className 所在类名
 * \@property description 节点自描述
 * @author linchang 2018/12/7
 */
FuncEntity = /** @class */ (function () {
    function FuncEntity() {
        this.className = "";
    }
    return FuncEntity;
}());
/**
 * 函数实体
 * \@property key 函数语义化名称
 * \@property params 函数参数集
 * \@property name 函数名称
 * \@property type 函数类型
 * \@property className 所在类名
 * \@property description 节点自描述
 * @author linchang 2018/12/7
 */
export { FuncEntity };
if (false) {
    /** @type {?} */
    FuncEntity.prototype.key;
    /** @type {?} */
    FuncEntity.prototype.params;
    /** @type {?} */
    FuncEntity.prototype.name;
    /** @type {?} */
    FuncEntity.prototype.type;
    /** @type {?} */
    FuncEntity.prototype.className;
    /** @type {?} */
    FuncEntity.prototype.description;
}
/**
 * 函数实体列表
 * \@property funcs 函数实体列表
 * @author linchang 2018/12/7
 */
var /**
 * 函数实体列表
 * \@property funcs 函数实体列表
 * @author linchang 2018/12/7
 */
FuncEntities = /** @class */ (function () {
    function FuncEntities() {
    }
    return FuncEntities;
}());
/**
 * 函数实体列表
 * \@property funcs 函数实体列表
 * @author linchang 2018/12/7
 */
export { FuncEntities };
if (false) {
    /** @type {?} */
    FuncEntities.prototype.funcs;
}
/** @enum {number} */
var FuncType = {
    "Advanced": 0,
    "Date": 1,
    "Estimate": 2,
    "String": 3,
    "Translate": 4,
    "Aggregate": 5,
    "Analytical": 6,
    "Math": 7,
    "Extend": 8 //扩展函数
    ,
};
export { FuncType };
FuncType[FuncType["Advanced"]] = "Advanced";
FuncType[FuncType["Date"]] = "Date";
FuncType[FuncType["Estimate"]] = "Estimate";
FuncType[FuncType["String"]] = "String";
FuncType[FuncType["Translate"]] = "Translate";
FuncType[FuncType["Aggregate"]] = "Aggregate";
FuncType[FuncType["Analytical"]] = "Analytical";
FuncType[FuncType["Math"]] = "Math";
FuncType[FuncType["Extend" //扩展函数
]] = "Extend" //扩展函数
;
/**
 * 参数信息
 * @param key 不同语言，统一标识
 * \@property name 语义化参数名
 * \@property type 参数类型
 * @author linchang 2018/12/7
 */
var /**
 * 参数信息
 * @param key 不同语言，统一标识
 * \@property name 语义化参数名
 * \@property type 参数类型
 * @author linchang 2018/12/7
 */
Parameter = /** @class */ (function () {
    function Parameter() {
    }
    return Parameter;
}());
/**
 * 参数信息
 * @param key 不同语言，统一标识
 * \@property name 语义化参数名
 * \@property type 参数类型
 * @author linchang 2018/12/7
 */
export { Parameter };
if (false) {
    /** @type {?} */
    Parameter.prototype.key;
    /** @type {?} */
    Parameter.prototype.name;
    /** @type {?} */
    Parameter.prototype.type;
}
/**
 * 参数信息
 * \@property key 语义化替换时脚本使用的统一主键key，一般为函数名，默认为英文
 * \@property name 脚本主键key对应的语义化字段，不同语言不同
 * \@property description 统一主键字段函数的功能描述
 * \@property type 函数类型
 * \@property keyFront 函数前缀（主要是DefaultFunction,Math）
 * \@property className 类路径（主要用户扩展函数寻找函数所在类路径）
 * @author manwenxing 2020/12/21
 */
var /**
 * 参数信息
 * \@property key 语义化替换时脚本使用的统一主键key，一般为函数名，默认为英文
 * \@property name 脚本主键key对应的语义化字段，不同语言不同
 * \@property description 统一主键字段函数的功能描述
 * \@property type 函数类型
 * \@property keyFront 函数前缀（主要是DefaultFunction,Math）
 * \@property className 类路径（主要用户扩展函数寻找函数所在类路径）
 * @author manwenxing 2020/12/21
 */
FunctionEntity = /** @class */ (function () {
    function FunctionEntity() {
    }
    return FunctionEntity;
}());
/**
 * 参数信息
 * \@property key 语义化替换时脚本使用的统一主键key，一般为函数名，默认为英文
 * \@property name 脚本主键key对应的语义化字段，不同语言不同
 * \@property description 统一主键字段函数的功能描述
 * \@property type 函数类型
 * \@property keyFront 函数前缀（主要是DefaultFunction,Math）
 * \@property className 类路径（主要用户扩展函数寻找函数所在类路径）
 * @author manwenxing 2020/12/21
 */
export { FunctionEntity };
if (false) {
    /** @type {?} */
    FunctionEntity.prototype.key;
    /** @type {?} */
    FunctionEntity.prototype.name;
    /** @type {?} */
    FunctionEntity.prototype.params;
    /** @type {?} */
    FunctionEntity.prototype.description;
    /** @type {?} */
    FunctionEntity.prototype.type;
    /** @type {?} */
    FunctionEntity.prototype.keyFront;
    /** @type {?} */
    FunctionEntity.prototype.className;
}
/** @type {?} */
export var FunctionEntitys = [
    {
        "key": "if",
        "name": "if()\n{\n\n}",
        "params": [],
        "description": "if判断条件",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "if else",
        "name": "if()\n{\n\n}\nelse\n{\n\n}",
        "params": [],
        "description": "if else判断条件",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "for",
        "name": "循环(;;)\n{\n\n}",
        "params": [],
        "description": "for循环",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "return",
        "name": "return ;",
        "params": [],
        "description": "返回操作",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "var",
        "name": "变量",
        "params": [],
        "description": "变量定义",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "true",
        "name": "true ",
        "params": [],
        "description": "正确",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "false",
        "name": "false ",
        "params": [],
        "description": "错误",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "StringSubstring",
        "name": "取子字符串",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "start",
                "name": "起始位置",
                "type": "int"
            },
            {
                "key": "length",
                "name": "长度",
                "type": "int"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5728\u6E90\u5B57\u7B26\u4E32\u4E2D\u68C0\u7D22\u5B50\u5B57\u7B26\u4E32\uFF0C\u5B50\u5B57\u7B26\u4E32\u4ECE\u6307\u5B9A\u7684\u5B57\u7B26\u4F4D\u7F6E\u5F00\u59CB\u4E14\u5177\u6709\u6307\u5B9A\u7684\u957F\u5EA6\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010start\u3011\uFF1A\u8D77\u59CB\u4F4D\u7F6E(\u4ECE0\u5F00\u59CB)\n\u3010length\u3011\uFF1A\u957F\u5EA6\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AStringSubstring(\"ABCDEFG\",2,3)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"CDE\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "SubString",
        "name": "取子字符串",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "start",
                "name": "起始位置",
                "type": "int"
            },
            {
                "key": "length",
                "name": "长度",
                "type": "int"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5728\u6E90\u5B57\u7B26\u4E32\u4E2D\u68C0\u7D22\u5B50\u5B57\u7B26\u4E32\uFF0C\u5B50\u5B57\u7B26\u4E32\u4ECE\u6307\u5B9A\u7684\u5B57\u7B26\u4F4D\u7F6E\u5F00\u59CB\u4E14\u5177\u6709\u6307\u5B9A\u7684\u957F\u5EA6\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010start\u3011\uFF1A\u8D77\u59CB\u4F4D\u7F6E(\u4ECE0\u5F00\u59CB)\n\u3010length\u3011\uFF1A\u957F\u5EA6\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ASubString(\"ABCDEFG\",2,3)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"CDE\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringLength",
        "name": "取字符串长度",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u83B7\u53D6\u6307\u5B9A\u5B57\u7B26\u4E32\u7684\u957F\u5EA6\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Int32)\n\u3010\u793A\u4F8B\u3011\uFF1AStringLength(\"\u8BED\u4E49\u5316\u8868\u8FBE\u5F0F\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A6",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Length",
        "name": "取字符串长度",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u83B7\u53D6\u6307\u5B9A\u5B57\u7B26\u4E32\u7684\u957F\u5EA6\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Int32)\n\u3010\u793A\u4F8B\u3011\uFF1ALength(\"\u8BED\u4E49\u5316\u8868\u8FBE\u5F0F\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A6",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Contains",
        "name": "是否包含",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5224\u65AD\u6E90\u5B57\u7B26\u4E32\u4E2D\u662F\u5426\u5305\u542B\u76EE\u6807\u5B57\u7B26\u4E32\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010target\u3011\uFF1A\u76EE\u6807\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1ASystem.Boolean\n\u3010\u793A\u4F8B\u3011\uFF1AContains(salesorder.code,\"\u8868\u8FBE\u5F0F\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1Atrue",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StartsWith",
        "name": "开头是",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5224\u65AD\u6E90\u5B57\u7B26\u4E32\u662F\u5426\u4EE5\u76EE\u6807\u5B57\u7B26\u4E32\u5F00\u5934\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010target\u3011\uFF1A\u76EE\u6807\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1ASystem.Boolean\n\u3010\u793A\u4F8B\u3011\uFF1AStartsWith(salesorder.code,\"\u5B50\u5355\u4F4D\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1Atrue",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "EndsWith",
        "name": "结尾是",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5224\u65AD\u6E90\u5B57\u7B26\u4E32\u662F\u5426\u4EE5\u76EE\u6807\u5B57\u7B26\u4E32\u7ED3\u5C3E\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010target\u3011\uFF1A\u76EE\u6807\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1ASystem.Boolean\n\u3010\u793A\u4F8B\u3011\uFF1AEndsWith(salesorder.code,\"\u5B50\u5355\u4F4D\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1Atrue",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringReplace",
        "name": "字符串替换",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "oldValue",
                "name": "被替换值",
                "type": "string"
            },
            {
                "key": "newValue",
                "name": "替换值",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6E90\u5B57\u7B26\u4E32\u4E2D\u7684\"\u88AB\u66FF\u6362\u503C\"\u66FF\u6362\u4E3A\"\u66FF\u6362\u503C\"\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010oldValue\u3011\uFF1A\u88AB\u66FF\u6362\u503C\n\u3010newValue\u3011\uFF1A\u66FF\u6362\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AStringReplace(\"123412\",\"12\",\"00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"003400\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Replace",
        "name": "字符串替换",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "oldValue",
                "name": "被替换值",
                "type": "string"
            },
            {
                "key": "newValue",
                "name": "替换值",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6E90\u5B57\u7B26\u4E32\u4E2D\u7684\"\u88AB\u66FF\u6362\u503C\"\u66FF\u6362\u4E3A\"\u66FF\u6362\u503C\"\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010oldValue\u3011\uFF1A\u88AB\u66FF\u6362\u503C\n\u3010newValue\u3011\uFF1A\u66FF\u6362\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AReplace(\"123412\",\"12\",\"00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"003400\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringToUpper",
        "name": "转大写",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6E90\u5B57\u7B26\u4E32\u8F6C\u6362\u4E3A\u5927\u5199\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AStringToUpper(\"abc\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"ABC\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToUpper",
        "name": "转大写",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6E90\u5B57\u7B26\u4E32\u8F6C\u6362\u4E3A\u5927\u5199\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AToUpper(\"abc\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"ABC\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringToLower",
        "name": "转小写",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6E90\u5B57\u7B26\u4E32\u8F6C\u6362\u4E3A\u5C0F\u5199\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AStringToLower(\"ABC\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"abc\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToLower",
        "name": "转小写",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6E90\u5B57\u7B26\u4E32\u8F6C\u6362\u4E3A\u5C0F\u5199\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AToLower(\"ABC\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"abc\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringIndexOf",
        "name": "获取首个匹配字符串的索引",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5728\u6E90\u5B57\u7B26\u4E32\u4E2D\uFF0C\u68C0\u7D22\u76EE\u6807\u5B57\u7B26\u4E32\uFF0C\u627E\u5230\u7B2C\u4E00\u4E2A\u5339\u914D\u7684\u5B57\u7B26\u4E32\u540E\uFF0C\u8FD4\u56DE\u5176\u7D22\u5F15\u4F4D\u7F6E(\u4ECE0\u5F00\u59CB)\u3002\u5982\u679C\u672A\u627E\u5230\u76EE\u6807\u5B57\u7B26\u4E32\uFF0C\u5219\u8FD4\u56DE-1\uFF0C\u5982\u679C\u76EE\u6807\u5B57\u7B26\u4E32\u4E3A\u7A7A\u4E32\uFF0C\u8FD4\u56DE0\uFF0C\u5982\u679C\u539F\u5B57\u7B26\u4E32\u4E3Anull\uFF0C\u5219\u8FD4\u56DE-1\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010target\u3011\uFF1A\u76EE\u6807\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Int32)\n\u3010\u793A\u4F8B\u3011\uFF1AStringIndexOf(\"12001234012\",\"01\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A3\n\u3010\u793A\u4F8B\u3011\uFF1AStringIndexOf(\"12002323\",\"34\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A-1\n\u3010\u793A\u4F8B\u3011\uFF1AStringIndexOf(\"120011\",\"\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A0",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "IndexOf",
        "name": "获取首个匹配字符串的索引",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5728\u6E90\u5B57\u7B26\u4E32\u4E2D\uFF0C\u68C0\u7D22\u76EE\u6807\u5B57\u7B26\u4E32\uFF0C\u627E\u5230\u7B2C\u4E00\u4E2A\u5339\u914D\u7684\u5B57\u7B26\u4E32\u540E\uFF0C\u8FD4\u56DE\u5176\u7D22\u5F15\u4F4D\u7F6E(\u4ECE0\u5F00\u59CB)\u3002\u5982\u679C\u672A\u627E\u5230\u76EE\u6807\u5B57\u7B26\u4E32\uFF0C\u5219\u8FD4\u56DE-1\uFF0C\u5982\u679C\u76EE\u6807\u5B57\u7B26\u4E32\u4E3A\u7A7A\u4E32\uFF0C\u8FD4\u56DE0\uFF0C\u5982\u679C\u539F\u5B57\u7B26\u4E32\u4E3Anull\uFF0C\u5219\u8FD4\u56DE-1\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010target\u3011\uFF1A\u76EE\u6807\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Int32)\n\u3010\u793A\u4F8B\u3011\uFF1AIndexOf(\"12001234012\",\"01\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A3\n\u3010\u793A\u4F8B\u3011\uFF1AIndexOf(\"12002323\",\"34\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A-1\n\u3010\u793A\u4F8B\u3011\uFF1AIndexOf(\"120011\",\"\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A0",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringLastIndexOf",
        "name": "获取最后一个匹配字符串的索引",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5728\u6E90\u5B57\u7B26\u4E32\u4E2D\uFF0C\u68C0\u7D22\u76EE\u6807\u5B57\u7B26\u4E32\uFF0C\u627E\u5230\u6700\u540E\u4E00\u4E2A\u5339\u914D\u7684\u5B57\u7B26\u4E32\u540E\uFF0C\u8FD4\u56DE\u5176\u7D22\u5F15\u4F4D\u7F6E(\u4ECE0\u5F00\u59CB)\u3002\u5982\u679C\u672A\u627E\u5230\u76EE\u6807\u5B57\u7B26\u4E32\uFF0C\u5219\u8FD4\u56DE-1\uFF0C\u5982\u679C\u76EE\u6807\u5B57\u7B26\u4E32\u4E3A\u7A7A\u4E32\uFF0C\u8FD4\u56DE0\uFF0C\u5982\u679C\u539F\u5B57\u7B26\u4E32\u4E3Anull\uFF0C\u5219\u8FD4\u56DE-1\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010value\u3011\uFF1A\u76EE\u6807\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Int32)\n\u3010\u793A\u4F8B\u3011\uFF1AStringLastIndexOf(\"12001234012\",\"01\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A8\n\u3010\u793A\u4F8B\u3011\uFF1AStringLastIndexOf(\"12002323\",\"34\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A-1\n\u3010\u793A\u4F8B\u3011\uFF1AStringLastIndexOf(\"120011\",\"\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A0",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastIndexOf",
        "name": "获取最后一个匹配字符串的索引",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5728\u6E90\u5B57\u7B26\u4E32\u4E2D\uFF0C\u68C0\u7D22\u76EE\u6807\u5B57\u7B26\u4E32\uFF0C\u627E\u5230\u6700\u540E\u4E00\u4E2A\u5339\u914D\u7684\u5B57\u7B26\u4E32\u540E\uFF0C\u8FD4\u56DE\u5176\u7D22\u5F15\u4F4D\u7F6E(\u4ECE0\u5F00\u59CB)\u3002\u5982\u679C\u672A\u627E\u5230\u76EE\u6807\u5B57\u7B26\u4E32\uFF0C\u5219\u8FD4\u56DE-1\uFF0C\u5982\u679C\u76EE\u6807\u5B57\u7B26\u4E32\u4E3A\u7A7A\u4E32\uFF0C\u8FD4\u56DE0\uFF0C\u5982\u679C\u539F\u5B57\u7B26\u4E32\u4E3Anull\uFF0C\u5219\u8FD4\u56DE-1\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010value\u3011\uFF1A\u76EE\u6807\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Int32)\n\u3010\u793A\u4F8B\u3011\uFF1ALastIndexOf(\"12001234012\",\"01\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A8\n\u3010\u793A\u4F8B\u3011\uFF1ALastIndexOf(\"12002323\",\"34\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A-1\n\u3010\u793A\u4F8B\u3011\uFF1ALastIndexOf(\"120011\",\"\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A0",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringTrimStart",
        "name": "删除头部空格",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5220\u9664\u6E90\u5B57\u7B26\u4E32\u7684\u5934\u90E8\u7A7A\u683C\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AStringTrimStart(\"    ABC\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"ABC\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "TrimStart",
        "name": "删除头部空格",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5220\u9664\u6E90\u5B57\u7B26\u4E32\u7684\u5934\u90E8\u7A7A\u683C\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ATrimStart(\"    ABC\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"ABC\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringTrimEnd",
        "name": "删除尾部空格",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5220\u9664\u6E90\u5B57\u7B26\u4E32\u7684\u5C3E\u90E8\u7A7A\u683C\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AStringTrimEnd(\"ABC    \")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"ABC\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "TrimEnd",
        "name": "删除尾部空格",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5220\u9664\u6E90\u5B57\u7B26\u4E32\u7684\u5C3E\u90E8\u7A7A\u683C\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ATrimEnd(\"ABC    \")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"ABC\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringTrim",
        "name": "删除头尾部空格",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5220\u9664\u6E90\u5B57\u7B26\u4E32\u7684\u5934\u5C3E\u90E8\u7A7A\u683C\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AStringTrim(\"    ABC    \")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"ABC\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Trim",
        "name": "删除头尾部空格",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5220\u9664\u6E90\u5B57\u7B26\u4E32\u7684\u5934\u5C3E\u90E8\u7A7A\u683C\n\u3010original\u3011\uFF1A\u6E90\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ATrim(\"    ABC    \")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"ABC\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "CreateGuid",
        "name": "唯一标识符",
        "params": [],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u83B7\u53D6\u552F\u4E00\u6807\u8BC6\u7B26Guid\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ACreateGuid()\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"7722A793-14DF-49CD-B7E7-DF4987YYD2\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringAdd",
        "name": "字符串相加",
        "params": [
            {
                "key": "firstStr",
                "name": "第一个字符串",
                "type": "string"
            },
            {
                "key": "secondStr",
                "name": "第二个字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u82E5\u5E72\u5B57\u7B26\u4E32\u62FC\u63A5\u6210\u4E00\u4E2A\u65B0\u5B57\u7B26\u4E32\uFF0C\u6700\u591A\u652F\u630110\u4E2A\u5B57\u7B26\u4E32\u62FC\u63A5\uFF0C\u6700\u5C112\u4E2A\uFF1B\u7F16\u8F91\u5668\u4E2D\u9ED8\u8BA4\u63D2\u5165\u4E24\u4E2A\uFF0C\u591A\u4E2A\u5B57\u7B26\u4E32\u53EF\u624B\u52A8\u6DFB\u52A0\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AStringAdd(\"Hello\",\"World\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"HelloWord\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Add",
        "name": "字符串相加",
        "params": [
            {
                "key": "firstStr",
                "name": "第一个字符串",
                "type": "string"
            },
            {
                "key": "secondStr",
                "name": "第二个字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u82E5\u5E72\u5B57\u7B26\u4E32\u62FC\u63A5\u6210\u4E00\u4E2A\u65B0\u5B57\u7B26\u4E32\uFF0C\u6700\u591A\u652F\u630110\u4E2A\u5B57\u7B26\u4E32\u62FC\u63A5\uFF0C\u6700\u5C112\u4E2A\uFF1B\u7F16\u8F91\u5668\u4E2D\u9ED8\u8BA4\u63D2\u5165\u4E24\u4E2A\uFF0C\u591A\u4E2A\u5B57\u7B26\u4E32\u53EF\u624B\u52A8\u6DFB\u52A0\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AAdd(\"Hello\",\"World\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"HelloWord\"",
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "SumByProp",
        "name": "按属性求和",
        "params": [
            {
                "key": "chainedPropName",
                "name": "实体集合属性名",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "类型属性名",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u7EDF\u8BA1\u67D0\u4E00\u5C5E\u6027\u7684\u548C\n\u3010chainedPropName\u3011\uFF1A\u5BF9\u5E94\u5B9E\u4F53\u96C6\u5408\u7684\u5C5E\u6027\u540D,\u94FE\u5F0F\u5F62\u5F0F\u7EC4\u7EC7\u5728\u4E00\u8D77\u7684\u5C5E\u6027\u540D,\u4F8B\u5982:Order.OrderItems\n\u3010propertyName\u3011\uFF1A\u7C7B\u578B\u5C5E\u6027\u540D;\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Double)\n\u3010\u793A\u4F8B\u3011\uFF1ASumByProp(\"Order.OrderItems\",\"Quantity\")\"",
        "type": "Aggregate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "CountByProp",
        "name": "列行数",
        "params": [
            {
                "key": "chainedPropName",
                "name": "实体集合属性名",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "类型属性名",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u7EDF\u8BA1\u6307\u5B9A\u5B57\u6BB5\u6240\u5728\u8868\u7684\u8BB0\u5F55\u6570\n\u3010chainedPropName\u3011\uFF1A\u5BF9\u5E94\u5B9E\u4F53\u96C6\u5408\u7684\u5C5E\u6027\u540D,\u94FE\u5F0F\u5F62\u5F0F\u7EC4\u7EC7\u5728\u4E00\u8D77\u7684\u5C5E\u6027\u540D,\u4F8B\u5982\uFF1AOrder.OrderItems\n\u3010propertyName\u3011\uFF1A\u7C7B\u578B\u5C5E\u6027\u540D\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Int32);\n\u3010\u793A\u4F8B\u3011\uFF1ACountByProp(\"Order.OrderItems\",\"Quantity\")",
        "type": "Aggregate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "AvgByProp",
        "name": "列平均值",
        "params": [
            {
                "key": "chainedPropName",
                "name": "实体集合属性名",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "类型属性名",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u7EDF\u8BA1\u67D0\u4E00\u6570\u503C\u5217\u7684\u5E73\u5747\u503C\n\u3010chainedPropName\u3011\uFF1A\u5BF9\u5E94\u5B9E\u4F53\u96C6\u5408\u7684\u5C5E\u6027\u540D,\u94FE\u5F0F\u5F62\u5F0F\u7EC4\u7EC7\u5728\u4E00\u8D77\u7684\u5C5E\u6027\u540D,\u4F8B\u5982:Order.OrderItems\n\u3010propertyName\u3011\uFF1A\u7C7B\u578B\u5C5E\u6027\u540D\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Double)\n\u3010\u793A\u4F8B\u3011\uFF1AAvgByProp(\"Order.OrderItems\",\"Quantity\")",
        "type": "Aggregate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "MaxByProp",
        "name": "列最大值",
        "params": [
            {
                "key": "chainedPropName",
                "name": "实体集合属性名",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "类型属性名",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u7EDF\u8BA1\u67D0\u4E00\u6570\u503C\u5217\u7684\u6700\u5927\u503C\n\u3010chainedPropName\u3011\uFF1A\u5BF9\u5E94\u5B9E\u4F53\u96C6\u5408\u7684\u5C5E\u6027\u540D,\u94FE\u5F0F\u5F62\u5F0F\u7EC4\u7EC7\u5728\u4E00\u8D77\u7684\u5C5E\u6027\u540D,\u4F8B\u5982:Order.OrderItems\n\u3010propertyName\u3011\uFF1A\u7C7B\u578B\u5C5E\u6027\u540D\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Double)\n\u3010\u793A\u4F8B\u3011\uFF1AMaxByProp(\"Order.OrderItems\",\"Quantity\")",
        "type": "Aggregate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "MinByProp",
        "name": "列最小值",
        "params": [
            {
                "key": "chainedPropName",
                "name": "实体集合属性名",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "类型属性名",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u7EDF\u8BA1\u67D0\u4E00\u6570\u503C\u5217\u7684\u6700\u5C0F\u503C\n\u3010chainedPropName\u3011\uFF1A\u5BF9\u5E94\u5B9E\u4F53\u96C6\u5408\u7684\u5C5E\u6027\u540D,\u94FE\u5F0F\u5F62\u5F0F\u7EC4\u7EC7\u5728\u4E00\u8D77\u7684\u5C5E\u6027\u540D,\u4F8B\u5982:Order.OrderItems\n\u3010propertyName\u3011\uFF1A\u7C7B\u578B\u5C5E\u6027\u540D\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Double)\n\u3010\u793A\u4F8B\u3011\uFF1AMinByProp(\"Order.OrderItems\",\"Quantity\")",
        "type": "Aggregate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "IsNull",
        "name": "是空值",
        "params": [
            {
                "key": "obj",
                "name": "检测对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5224\u65AD\u6307\u5B9A\u7684\u5BF9\u8C61\u662F\u5426\u4E3A\u7A7A\u503C(Null)\n\u3010obj\u3011\uFF1A\u8981\u68C0\u6D4B\u7684\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5E03\u5C14\u503C(System.Boolean)\n\u3010\u793A\u4F8B\u3011\uFF1AIsNull(\"\u8BED\u4E49\u5316\u8868\u8FBE\u5F0F\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1AFalse",
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "IsNullOrWhiteSpace",
        "name": "是空串",
        "params": [
            {
                "key": "obj",
                "name": "待判断字符串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5224\u65AD\u6307\u5B9A\u7684\u5B57\u7B26\u4E32\u662F\u5426\u4E3Anull\u3001\u7A7A\u8FD8\u662F\u4EC5\u6709\u7A7A\u683C\u7EC4\u6210\n\u3010obj\u3011\uFF1A\u8981\u5224\u65AD\u7684\u5B57\u7B26\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5E03\u5C14\u503C(System.Boolean)\n\u3010\u793A\u4F8B\u3011\uFF1AIsNullOrWhiteSpace(\"\u8BED\u4E49\u5316\u8868\u8FBE\u5F0F\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1AFalse\n\u3010\u793A\u4F8B\u3011\uFF1AIsNullOrWhiteSpace(null)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1ATrue\n\u3010\u793A\u4F8B\u3011\uFF1A\nIsNullOrWhiteSpace(\" \")\u3010\u8FD4\u56DE\u503C\u3011\uFF1ATrue",
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "IsNaN",
        "name": "非数字",
        "params": [
            {
                "key": "obj",
                "name": "检测对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5224\u65AD\u6307\u5B9A\u7684\u5BF9\u8C61\u662F\u5426\u53EF\u4EE5\u8F6C\u6362\u6210\u6570\u5B57(System.Double)\n\u3010obj\u3011\uFF1A\u8981\u68C0\u6D4B\u7684\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5E03\u5C14\u503C(System.Boolean)\n\u3010\u793A\u4F8B\u3011\uFF1AIsNaN(\"\u8BED\u4E49\u5316\u8868\u8FBE\u5F0F\")\u3010\u8FD4\u56DE\u503C\u3011\uFF1ATrue\n\u3010\u793A\u4F8B\u3011\uFF1AIsNaN(12.3)\u3010\u8FD4\u56DE\u503C\u3011\uFF1AFalse\n\u3010\u793A\u4F8B\u3011\uFF1A\nIsNaN(\"123\")\u3010\u8FD4\u56DE\u503C\u3011\uFF1AFalse",
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "IsNumber",
        "name": "是数字",
        "params": [
            {
                "key": "obj",
                "name": "检测对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5224\u65AD\u6307\u5B9A\u7684\u5BF9\u8C61\u662F\u5426\u53EF\u4EE5\u8F6C\u6362\u6210\u6570\u5B57(System.Double)\n\u3010obj\u3011\uFF1A\u8981\u68C0\u6D4B\u7684\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5E03\u5C14\u503C(System.Boolean)\n\u3010\u793A\u4F8B\u3011\uFF1AIsNumber(\"\u8BED\u4E49\u5316\u8868\u8FBE\u5F0F\")\u3010\u8FD4\u56DE\u503C\u3011\uFF1AFalse\n\u3010\u793A\u4F8B\u3011\uFF1AIsNumber(12.3)\u3010\u8FD4\u56DE\u503C\u3011\uFF1ATrue\n\u3010\u793A\u4F8B\u3011\uFF1A\nIsNumber(\"123\")\u3010\u8FD4\u56DE\u503C\u3011\uFF1ATrue",
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "IsExistRecord",
        "name": "有无等于匹配字符",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            },
            {
                "key": "match",
                "name": "匹配值",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u6839\u636E\u94FE\u5F0F\u540D\u79F0\u5224\u65AD\u8BE5\u8868\u7684\u6307\u5B9A\u5B57\u6BB5\u96C6\u5408\u4E2D\u662F\u5426\u5B58\u5728\u7B49\u4E8E\u5339\u914D\u5B57\u7B26\u7684\u503C\n\u3010chainedPropName\u3011\u94FE\u5F0F\u540D\u79F0\n\u3010propertyName\u3011\u5C5E\u6027\u540D\n\u3010match\u3011\u5339\u914D\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5BF9\u8C61(System.Boolean)\n\u3010\u793A\u4F8B\u3011\uFF1AIsExistRecord(\"salesorder.items\",\"code\",\"lc\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1AFalse/True",
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ListGreaterThan",
        "name": "有无大于匹配字符",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            },
            {
                "key": "match",
                "name": "匹配值",
                "type": "double"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u6839\u636E\u94FE\u5F0F\u540D\u79F0\u5224\u65AD\u8BE5\u8868\u7684\u6307\u5B9A\u5B57\u6BB5\u96C6\u5408\u4E2D\u662F\u5426\u5B58\u5728\u5927\u4E8E\u5339\u914D\u5B57\u7B26\u7684\u503C\n\u3010chainedPropName\u3011\u94FE\u5F0F\u540D\u79F0\n\u3010propertyName\u3011\u5C5E\u6027\u540D\n\u3010match\u3011\u5339\u914D\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5BF9\u8C61(System.Boolean)\n\u3010\u793A\u4F8B\u3011\uFF1AListGreaterThan(\"salesorder.items\",\"num\",100)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1AFalse/True",
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ListLessThan",
        "name": "有无小于匹配字符",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            },
            {
                "key": "match",
                "name": "匹配值",
                "type": "double"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u6839\u636E\u94FE\u5F0F\u540D\u79F0\u5224\u65AD\u8BE5\u8868\u7684\u6307\u5B9A\u5B57\u6BB5\u96C6\u5408\u4E2D\u662F\u5426\u5B58\u5728\u5C0F\u4E8E\u5339\u914D\u5B57\u7B26\u7684\u503C\n\u3010chainedPropName\u3011\u94FE\u5F0F\u540D\u79F0\n\u3010propertyName\u3011\u5C5E\u6027\u540D\n\u3010match\u3011\u5339\u914D\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5BF9\u8C61(System.Boolean)\n\u3010\u793A\u4F8B\u3011\uFF1AListLessThan(\"salesorder.items\",\"num\",100)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1AFalse/True",
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ListStartWith",
        "name": "有无开头是匹配字符",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            },
            {
                "key": "match",
                "name": "匹配值",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u6839\u636E\u94FE\u5F0F\u540D\u79F0\u5224\u65AD\u8BE5\u8868\u7684\u6307\u5B9A\u5B57\u6BB5\u96C6\u5408\u4E2D\u662F\u5426\u5B58\u5728\u5F00\u5934\u662F\u5339\u914D\u5B57\u7B26\u7684\u503C\n\u3010chainedPropName\u3011\u94FE\u5F0F\u540D\u79F0\n\u3010propertyName\u3011\u5C5E\u6027\u540D\n\u3010match\u3011\u5339\u914D\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5BF9\u8C61(System.Boolean)\n\u3010\u793A\u4F8B\u3011\uFF1AListStartWith(\"salesorder.items\",\"code\",\"lc\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1AFalse/True",
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ListEndWith",
        "name": "有无结尾是匹配字符",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            },
            {
                "key": "match",
                "name": "匹配值",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u6839\u636E\u94FE\u5F0F\u540D\u79F0\u5224\u65AD\u8BE5\u8868\u7684\u6307\u5B9A\u5B57\u6BB5\u96C6\u5408\u4E2D\u662F\u5426\u5B58\u5728\u7ED3\u5C3E\u662F\u5339\u914D\u5B57\u7B26\u7684\u503C\n\u3010chainedPropName\u3011\u94FE\u5F0F\u540D\u79F0\n\u3010propertyName\u3011\u5C5E\u6027\u540D\n\u3010match\u3011\u5339\u914D\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5BF9\u8C61(System.Boolean)\n\u3010\u793A\u4F8B\u3011\uFF1AListEndWith(\"salesorder.items\",\"code\",\"lc\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1AFalse/True",
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "DateTimeAddDays",
        "name": "增加指定天数",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            },
            {
                "key": "value",
                "name": "增加天数",
                "type": "int"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5728\u6307\u5B9A\u7684\u65E5\u671F\u65F6\u95F4\u4E0A\u589E\u52A0\u6307\u5B9A\u7684\u5929\u6570\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010value\u3011\uFF1A\u589E\u52A0\u5929\u6570\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u65E5\u671F\u65F6\u95F4(System.DateTime)\n\u3010\u793A\u4F8B\u3011\uFF1ADateTimeAddDays(\"2019-01-30T10:00:00+08:00\",2)\n\u3010\u8FD4\u56DE\u503C\u3011:2019-02-01T10:00+08:00",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "DateTimeAddYears",
        "name": "增加指定年数",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            },
            {
                "key": "value",
                "name": "增加年数",
                "type": "int"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5728\u6307\u5B9A\u7684\u65E5\u671F\u65F6\u95F4\u4E0A\u589E\u52A0\u6307\u5B9A\u7684\u5E74\u6570\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010value\u3011\uFF1A\u589E\u52A0\u5E74\u6570\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u65E5\u671F\u65F6\u95F4(System.DateTime)\n\u3010\u793A\u4F8B\u3011\uFF1ADateTimeAddYears(\"2019-01-30T10:00:00+08:00\",2)\n\u3010\u8FD4\u56DE\u503C\u3011:2021-01-30T10:00+08:00",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "DateTimeAddMonths",
        "name": "增加指定月数",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            },
            {
                "key": "value",
                "name": "增加月数",
                "type": "int"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5728\u6307\u5B9A\u7684\u65E5\u671F\u65F6\u95F4\u4E0A\u589E\u52A0\u6307\u5B9A\u7684\u6708\u6570\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010value\u3011\uFF1A\u589E\u52A0\u6708\u6570\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u65E5\u671F\u65F6\u95F4(System.DateTime)\n\u3010\u793A\u4F8B\u3011\uFF1ADateTimeAddMonths(\"2019-01-30T10:00:00+08:00\",1)\n\u3010\u8FD4\u56DE\u503C\u3011:2019-02-28T10:00+08:00",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetFirstDayOfMonth",
        "name": "获取指定日期时间所在月第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u83B7\u53D6\u6307\u5B9A\u65E5\u671F\u65F6\u95F4\u503C\u6240\u5728\u6708\u7684\u7B2C\u4E00\u5929\u7684\u65E5\u671F\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u65E5\u671F\u65F6\u95F4(System.DateTime)\n\u3010\u793A\u4F8B\u3011\uFF1AGetFirstDayOfMonth(\"2019-01-30T10:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011:2019-01-01T10:00+08:00",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetLastDayOfMonth",
        "name": "获取指定日期时间所在月最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u83B7\u53D6\u6307\u5B9A\u65E5\u671F\u65F6\u95F4\u503C\u6240\u5728\u6708\u7684\u6700\u540E\u4E00\u5929\u7684\u65E5\u671F\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u65E5\u671F\u65F6\u95F4(System.DateTime)\n\u3010\u793A\u4F8B\u3011\uFF1AGetLastDayOfMonth(\"2019-01-30T10:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011:2019-01-31T10:00+08:00",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetDateTimeNow",
        "name": "获取当前日期时间",
        "params": [],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u83B7\u53D6\u5F53\u524D\u7684\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u65E5\u671F\u65F6\u95F4(System.DateTime)\n\u3010\u793A\u4F8B\u3011\uFF1AGetDateTimeNow()\n\u3010\u8FD4\u56DE\u503C\u3011:2019-01-30T12:00:00.000+08:00",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "CompareDate",
        "name": "比较两个时间",
        "params": [
            {
                "key": "str1",
                "name": "日期时间1",
                "type": "string"
            },
            {
                "key": "str2",
                "name": "日期时间2",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u6BD4\u8F83\u4E24\u4E2A\u65F6\u95F4\u503C\uFF0C\u5E76\u8FD4\u56DE\u4E00\u4E2A\u503C\uFF0C\u6B64\u503C\u8868\u793A\u7B2C\u4E00\u4E2A\u65F6\u95F4\u662F\u65E9\u4E8E\u3001\u7B49\u4E8E\u8FD8\u662F\u665A\u4E8E\u7B2C\u4E8C\u4E2A\u65F6\u95F4\u7684\u6574\u6570\n\u3010str1\u3011\uFF1A\u65E5\u671F\u65F6\u95F41\n\u3010str2\u3011\uFF1A\u65E5\u671F\u65F6\u95F42\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.int32)\n\t\u65E5\u671F1\u65E9\u4E8E\u65E5\u671F2\uFF0C\u8FD4\u56DE-1\n\t\u65E5\u671F1\u7B49\u4E8E\u65E5\u671F2\uFF0C\u8FD4\u56DE0\n\t\u65E5\u671F1\u665A\u4E8E\u65E5\u671F2\uFF0C\u8FD4\u56DE1\n\u3010\u793A\u4F8B\u3011\uFF1ACompareDate(\"2019-01-30T10:00:00+08:00\",\"2019-02-01T10:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A-1",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FormatDate",
        "name": "格式化当前日期",
        "params": [
            {
                "key": "format",
                "name": "格式串",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u4F7F\u7528\u6307\u5B9A\u7684\u683C\u5F0F\u5C06\u5F53\u524D\u65E5\u671F\u65F6\u95F4\u503C\u8F6C\u6362\u4E3A\u5176\u7B49\u6548\u7684\u5B57\u7B26\u4E32\u8868\u793A\u5F62\u5F0F\n\u3010format\u3011\uFF1A\u683C\u5F0F\u4E32\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AFormatDate(\"yyyy-MM-dd HH:mm:ss\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-01-30 01:00:00\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FormatDefineDate",
        "name": "格式化指定日期",
        "params": [
            {
                "key": "format",
                "name": "格式串",
                "type": "string"
            },
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u4F7F\u7528\u6307\u5B9A\u7684\u683C\u5F0F\u5C06\u6307\u5B9A\u7684\u65E5\u671F\u65F6\u95F4\u503C\u8F6C\u6362\u4E3A\u5176\u7B49\u6548\u7684\u5B57\u7B26\u4E32\u8868\u793A\u5F62\u5F0F\n\u3010format\u3011\uFF1A\u683C\u5F0F\u4E32\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AFormatDefineDate(\"yyyy-MM-dd HH:mm:ss\",\"2019-01-30 01:00:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-01-30 01:00:00\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Today",
        "name": "今天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4ECA\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\u5F62\u5F0F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4ECA\u5929\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AToday(\"2019-01-30T10:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-01-30\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Yesterday",
        "name": "昨天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u6628\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\u5F62\u5F0F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u6628\u5929\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AYesterday(\"2019-01-30T10:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-01-29\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Tomorrow",
        "name": "明天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u660E\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\u5F62\u5F0F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u660E\u5929\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ATomorrow(\"2019-01-30T10:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-01-31\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ThisYear",
        "name": "本年",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u672C\u5E74\uFF0C\u683C\u5F0F\u4E3A\"yyyy\"\u5F62\u5F0F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u672C\u5E74\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AThisYear(\"2019-01-30T10:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastYear",
        "name": "去年",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u53BB\u5E74\uFF0C\u683C\u5F0F\u4E3A\"yyyy\"\u5F62\u5F0F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u53BB\u5E74\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ALastYear(\"2019-01-30T10:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2018\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "NextYear",
        "name": "明年",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u660E\u5E74\uFF0C\u683C\u5F0F\u4E3A\"yyyy\"\u5F62\u5F0F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u660E\u5E74\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ANextYear(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2020\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ThisMonth",
        "name": "本月",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u672C\u6708\uFF0C\u683C\u5F0F\u4E3A\"MM\"\u5F62\u5F0F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u672C\u6708\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AThisMonth(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"01\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastMonth",
        "name": "上月",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0A\u6708\uFF0C\u683C\u5F0F\u4E3A\"MM\"\u5F62\u5F0F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0A\u6708\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ALastMonth(\"2019-02-30 01:00:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"01\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "NextMonth",
        "name": "下月",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0B\u6708\uFF0C\u683C\u5F0F\u4E3A\"MM\"\u5F62\u5F0F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0B\u6708\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ANextMonth(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"02\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ThisWeek",
        "name": "本周",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u5728\u672C\u5E74\u7684\u7B2C\u51E0\u5468\uFF0C\u683C\u5F0F\u4E3A\"ww\"\u5F62\u5F0F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u5728\u672C\u5E74\u7684\u7B2C\u51E0\u5468\uFF0C\u6CE8\u610F\u5982\u679C12\u6708\u5E95\u7684\u65E5\u671F\u4E0E\u4E0B\u4E00\u5E74\u76841\u67081\u65E5\u5728\u540C\u4E00\u661F\u671F\uFF0C\u5219\u8FD4\u56DE\u7B2C\u4E00\u5468\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AThisWeek(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"5\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastWeek",
        "name": "上周",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0A\u5468\uFF0C\u683C\u5F0F\u4E3A\"ww\"\u5F62\u5F0F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0A\u5468\uFF0C\u6CE8\u610F\u5982\u679C\u6307\u5B9A\u65E5\u671F\u5728\u4ECA\u5E74\u7B2C1\u5468\uFF0C\u5219\u4E0A\u5468\u4E3A\u53BB\u5E74\u7684\u6700\u540E\u4E00\u5468\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ALastWeek(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"4\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "NextWeek",
        "name": "下周",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0B\u5468\uFF0C\u683C\u5F0F\u4E3A\"ww\"\u5F62\u5F0F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0B\u5468\uFF0C\u6CE8\u610F\u5982\u679C\u6307\u5B9A\u65E5\u671F\u5728\u4ECA\u5E74\u6700\u540E\u4E00\u5468\uFF0C\u5219\u4E0B\u5468\u4E3A1\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ANextWeek(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"6\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "DayOfWeek",
        "name": "星期几",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u662F\u672C\u5468\u7684\u661F\u671F\u51E0\uFF0C\u683C\u5F0F\u4E3A\"n\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u662F\u672C\u5468\u7684\u661F\u671F\u51E0\uFF0C\u6CE8\u610F\u4E0D\u662F\u8FD4\u56DE\u82F1\u6587\u7F29\u5199\uFF0C\u800C\u662F\u8FD4\u56DE\u6570\u5B57\u5F62\u5F0F\u3002\u6309\u7167\u4E2D\u56FD\u4E60\u60EF\uFF0C\u661F\u671F\u4E00\u4E3A\u6BCF\u5468\u9996\u65E5\uFF0C\u8FD4\u56DE\u503C\u4E3A1\uFF0C\u661F\u671F\u65E5\u4E3A\u6BCF\u5468\u5C3E\u65E5\uFF0C\u8FD4\u56DE\u503C\u4E3A7\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ADayOfWeek(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"3\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfWeek",
        "name": "本周第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u672C\u5468\u7B2C\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u672C\u5468\u7B2C\u4E00\u5929\u3002\u6309\u7167\u4E2D\u56FD\u4E60\u60EF\uFF0C\u661F\u671F\u4E00\u4E3A\u6BCF\u5468\u9996\u65E5\uFF0C\u661F\u671F\u65E5\u4E3A\u6BCF\u5468\u5C3E\u65E5\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AFirstDayOfWeek(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-01-28\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfWeek",
        "name": "本周最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u672C\u5468\u6700\u540E\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u672C\u5468\u7B2C\u4E00\u5929\u3002\u6309\u7167\u4E2D\u56FD\u4E60\u60EF\uFF0C\u661F\u671F\u4E00\u4E3A\u6BCF\u5468\u9996\u65E5\uFF0C\u661F\u671F\u65E5\u4E3A\u6BCF\u5468\u5C3E\u65E5\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AFirstDayOfWeek(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-02-03\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfLastWeek",
        "name": "上周第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0A\u5468\u7B2C\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0A\u5468\u7B2C\u4E00\u5929\u3002\u6309\u7167\u4E2D\u56FD\u4E60\u60EF\uFF0C\u661F\u671F\u4E00\u4E3A\u6BCF\u5468\u9996\u65E5\uFF0C\u661F\u671F\u65E5\u4E3A\u6BCF\u5468\u5C3E\u65E5\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AFirstDayOfLastWeek(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-01-21\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfLastWeek",
        "name": "上周最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0A\u5468\u6700\u540E\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0A\u5468\u6700\u540E\u4E00\u5929\u3002\u6309\u7167\u4E2D\u56FD\u4E60\u60EF\uFF0C\u661F\u671F\u4E00\u4E3A\u6BCF\u5468\u9996\u65E5\uFF0C\u661F\u671F\u65E5\u4E3A\u6BCF\u5468\u5C3E\u65E5\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ALastDayOfLastWeek(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-01-27\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfNextWeek",
        "name": "下周第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0B\u5468\u7B2C\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0B\u5468\u7B2C\u4E00\u5929\u3002\u6309\u7167\u4E2D\u56FD\u4E60\u60EF\uFF0C\u661F\u671F\u4E00\u4E3A\u6BCF\u5468\u9996\u65E5\uFF0C\u661F\u671F\u65E5\u4E3A\u6BCF\u5468\u5C3E\u65E5\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AFirstDayOfNextWeek(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-02-04\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfNextWeek",
        "name": "下周最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0B\u5468\u6700\u540E\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0B\u5468\u6700\u540E\u4E00\u5929\u3002\u6309\u7167\u4E2D\u56FD\u4E60\u60EF\uFF0C\u661F\u671F\u4E00\u4E3A\u6BCF\u5468\u9996\u65E5\uFF0C\u661F\u671F\u65E5\u4E3A\u6BCF\u5468\u5C3E\u65E5\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ALastDayOfNextWeek(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-02-10\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfMonth",
        "name": "本月第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u672C\u6708\u7B2C\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u672C\u6708\u7B2C\u4E00\u5929\u3002\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AFirstDayOfMonth(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-01-01\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfMonth",
        "name": "本月最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u672C\u6708\u6700\u540E\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u672C\u6708\u6700\u540E\u4E00\u5929\u3002\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ALastDayOfMonth(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-01-31\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfLastMonth",
        "name": "上月第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0A\u6708\u7B2C\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0A\u6708\u7B2C\u4E00\u5929\u3002\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AFirstDayOfLastMonth(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2018-12-01\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfLastMonth",
        "name": "上月最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0A\u6708\u6700\u540E\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0A\u6708\u6700\u540E\u4E00\u5929\u3002\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ALastDayOfLastMonth(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2018-12-31\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfNextMonth",
        "name": "下月第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0B\u6708\u7B2C\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0B\u6708\u7B2C\u4E00\u5929\u3002\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AFirstDayOfNextMonth(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-02-01\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfNextMonth",
        "name": "下月最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0B\u6708\u6700\u540E\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0B\u6708\u6700\u540E\u4E00\u5929\u3002\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ALastDayOfNextMonth(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-02-28\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfYear",
        "name": "本年第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u672C\u5E74\u7B2C\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u672C\u5E74\u7B2C\u4E00\u5929\u3002\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AFirstDayOfYear(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-01-01\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfYear",
        "name": "本年最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u672C\u5E74\u6700\u540E\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u672C\u5E74\u6700\u540E\u4E00\u5929\u3002\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ALastDayOfYear(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-12-31\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfLastYear",
        "name": "上年第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0A\u5E74\u7B2C\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0A\u5E74\u7B2C\u4E00\u5929\u3002\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AFirstDayOfLastYear(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2018-01-01\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfLastYear",
        "name": "上年最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0A\u5E74\u6700\u540E\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0A\u5E74\u6700\u540E\u4E00\u5929\u3002\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ALastDayOfLastYear(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2018-12-31\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfNextYear",
        "name": "下年第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0B\u5E74\u7B2C\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0B\u5E74\u7B2C\u4E00\u5929\u3002\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AFirstDayOfNextYear(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2020-01-01\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfNextYear",
        "name": "下年最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u65E5\u671F\u7684\u4E0B\u5E74\u6700\u540E\u4E00\u5929\uFF0C\u683C\u5F0F\u4E3A\"yyyy-MM-dd\"\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u4E0B\u5E74\u6700\u540E\u4E00\u5929\u3002\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1ALastDayOfNextYear(\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2020-12-31\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": "DefaultFunction"
    },
    {
        "key": "GetDate",
        "name": "返回指定形式的日期",
        "params": [
            {
                "key": "format",
                "name": "格式串",
                "type": "string"
            },
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u6307\u5B9A\u5F62\u5F0F\u7684\u65E5\u671F\uFF0C\u5982\u679C\u4E0D\u4F20\u9012\u53C2\u6570\uFF0C\u5219\u9ED8\u8BA4\u8FD4\u56DE\u5F53\u524D\u7CFB\u7EDF\u65E5\u671F\u7684\u6307\u5B9A\u5F62\u5F0F\u65E5\u671F\u3002\n\u3010format\u3011\uFF1A\u683C\u5F0F\u5316\u4E32\n\u3010date\u3011\uFF1A\u65E5\u671F\u65F6\u95F4\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\tDD\uFF1A\u4ECA\u5929\tLD\uFF1A\u6628\u5929\tND\uFF1A\u660E\u5929\n\tYY\uFF1A\u672C\u5E74\tLY\uFF1A\u4E0A\u5E74\tNY\uFF1A\u4E0B\u5E74\n\tMM\uFF1A\u672C\u6708\tLM\uFF1A\u4E0A\u6708\tNM\uFF1A\u4E0B\u6708\n\tWD\uFF1A\u661F\u671F\u51E0\n\tWW\uFF1A\u672C\u5468\tLW\uFF1A\u4E0A\u5468\tNW\uFF1A\u4E0B\u5468\n\tFDW\uFF1A\u672C\u5468\u9996\u65E5\tLDW\uFF1A\u672C\u5468\u5C3E\u65E5\n\tFDLW\uFF1A\u4E0A\u5468\u9996\u65E5\tLDLW\uFF1A\u4E0A\u5468\u5C3E\u65E5\n\tFDNW\uFF1A\u4E0B\u5468\u9996\u65E5\tLDNW\uFF1A\u4E0B\u5468\u5C3E\u65E5\n\tFDM\uFF1A\u672C\u6708\u9996\u65E5\tLDM\uFF1A\u672C\u6708\u5C3E\u65E5\n\tFDLM\uFF1A\u4E0A\u6708\u9996\u65E5\tLDLM\uFF1A\u4E0A\u6708\u5C3E\u65E5\n\tFDNM\uFF1A\u4E0B\u6708\u9996\u65E5\tLDNM\uFF1A\u4E0B\u6708\u5C3E\u65E5\n\tFDY\uFF1A\u672C\u5E74\u9996\u65E5\tLDY\uFF1A\u672C\u5E74\u5C3E\u65E5\n\tFDLY\uFF1A\u4E0A\u5E74\u9996\u65E5\tLDLY\uFF1A\u4E0A\u5E74\u5C3E\u65E5\n\tFDNY\uFF1A\u4E0B\u5E74\u9996\u65E5\tLDNY\uFF1A\u4E0B\u5E74\u5C3E\u65E5\n\u3010\u793A\u4F8B\u3011\uFF1AGetDate(\"LD\",\"2019-01-30T01:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"2019-01-29\"",
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToStringX",
        "name": "转字符串",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6307\u5B9A\u7684\u5BF9\u8C61\u8F6C\u6362\u6210\u5B57\u7B26\u4E32\n\u3010obj\u3011\uFF1A\u5F85\u8F6C\u6362\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AToStringX(123)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"123\"",
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToDateTime",
        "name": "转日期时间",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6307\u5B9A\u7684\u5BF9\u8C61\u8F6C\u6362\u6210\u65E5\u671F\u65F6\u95F4\n\u3010obj\u3011\uFF1A\u5F85\u8F6C\u6362\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u65E5\u671F\u65F6\u95F4(System.OffsetDateTime)\n\u3010\u793A\u4F8B\u3011\uFF1AToDateTime(\"2019-01-30T10:00:00+08:00\")\n\u3010\u8FD4\u56DE\u503C\u3011: 2019-01-30T10:00+08:00",
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToBoolean",
        "name": "转布尔值",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6307\u5B9A\u5BF9\u8C61\u503C\u8F6C\u6362\u6210\u7B49\u6548\u7684\u5E03\u5C14\u503C\n\u3010obj\u3011\uFF1A\u5F85\u8F6C\u6362\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5E03\u5C14\u503C(System.Boolean)\n\u3010\u793A\u4F8B\u3011\uFF1AToBoolean(\"True\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1Atrue",
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToDecimal",
        "name": "转数值",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6307\u5B9A\u5BF9\u8C61\u8F6C\u6362\u6210\u5341\u8FDB\u5236\u6570\u503C\n\u3010obj\u3011\uFF1A\u5F85\u8F6C\u6362\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5341\u8FDB\u5236\u6570\u503C(System.BigDecimal)\n\u3010\u793A\u4F8B\u3011\uFF1AToDecimal(\"40.23\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A40.23",
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToBigDecimal",
        "name": "转数值",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6307\u5B9A\u5BF9\u8C61\u8F6C\u6362\u6210\u5341\u8FDB\u5236\u6570\u503C\n\u3010obj\u3011\uFF1A\u5F85\u8F6C\u6362\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5341\u8FDB\u5236\u6570\u503C(System.BigDecimal)\n\u3010\u793A\u4F8B\u3011\uFF1AToBigDecimal(\"40.23\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A40.23",
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToDouble",
        "name": "转双精度浮点数",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6307\u5B9A\u5BF9\u8C61\u8F6C\u6362\u6210\u53CC\u7CBE\u5EA6\u6D6E\u70B9\u6570\n\u3010obj\u3011\uFF1A\u5F85\u8F6C\u6362\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u53CC\u7CBE\u5EA6\u6D6E\u70B9\u6570(System.Double)\n\u3010\u793A\u4F8B\u3011\uFF1AToDouble(\"40.23\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A40.23",
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToSingle",
        "name": "转单精度浮点数",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6307\u5B9A\u5BF9\u8C61\u8F6C\u6362\u6210\u5355\u7CBE\u5EA6\u6D6E\u70B9\u6570\n\u3010obj\u3011\uFF1A\u5F85\u8F6C\u6362\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5355\u7CBE\u5EA6\u6D6E\u70B9\u6570(System.Float)\n\u3010\u793A\u4F8B\u3011\uFF1AToSingle(\"40.23\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A40.23",
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToFloat",
        "name": "转单精度浮点数",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6307\u5B9A\u5BF9\u8C61\u8F6C\u6362\u6210\u5355\u7CBE\u5EA6\u6D6E\u70B9\u6570\n\u3010obj\u3011\uFF1A\u5F85\u8F6C\u6362\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5355\u7CBE\u5EA6\u6D6E\u70B9\u6570(System.Float)\n\u3010\u793A\u4F8B\u3011\uFF1AToFloat(\"40.23\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A40.23",
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToByte",
        "name": "转8位无符号整数",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6307\u5B9A\u5BF9\u8C61\u8F6C\u6362\u62108\u4F4D\u65E0\u7B26\u53F7\u6574\u6570\n\u3010obj\u3011\uFF1A\u5F85\u8F6C\u6362\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A8\u4F4D\u65E0\u7B26\u53F7\u6574\u6570(System.Byte)\n\u3010\u793A\u4F8B\u3011\uFF1AToByte(\"10\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A10",
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToInt16",
        "name": "转16位有符号整数",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6307\u5B9A\u5BF9\u8C61\u8F6C\u6362\u621016\u4F4D\u6709\u7B26\u53F7\u6574\u6570\n\u3010obj\u3011\uFF1A\u5F85\u8F6C\u6362\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A16\u4F4D\u6709\u7B26\u53F7\u6574\u6570(System.Int16)\n\u3010\u793A\u4F8B\u3011\uFF1AToInt16(\"10\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A10",
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToInt32",
        "name": "转32位有符号整数",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6307\u5B9A\u5BF9\u8C61\u8F6C\u6362\u621032\u4F4D\u6709\u7B26\u53F7\u6574\u6570\n\u3010obj\u3011\uFF1A\u5F85\u8F6C\u6362\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A32\u4F4D\u6709\u7B26\u53F7\u6574\u6570(System.Int32)\n\u3010\u793A\u4F8B\u3011\uFF1AToInt32(\"10\")\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A10",
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToChineseMoney",
        "name": "转中文大写金额",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u5C06\u6307\u5B9A\u5BF9\u8C61\u8F6C\u6362\u6210\u4E2D\u6587\u5927\u5199\u91D1\u989D\n\u3010obj\u3011\uFF1A\u5F85\u8F6C\u6362\u5BF9\u8C61\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5B57\u7B26\u4E32(System.String)\n\u3010\u793A\u4F8B\u3011\uFF1AToChineseMoney(40.23)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A\"\u8086\u62FE\u5143\u96F6\u8D30\u89D2\u53C1\u5206\"",
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetSessionValue",
        "name": "获取会话变量的值",
        "params": [
            {
                "key": "name",
                "name": "会话变量名",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u83B7\u53D6\u7528\u6237\u4F1A\u8BDD\u53D8\u91CF\u7684\u503C\n\u3010name\u3011\uFF1A\u4F1A\u8BDD\u53D8\u91CF\u540D\u79F0\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5BF9\u8C61(System.Object)",
        "type": "Analytical",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetContextParameter",
        "name": "获取上下文变量的值",
        "params": [
            {
                "key": "name",
                "name": "上下文变量名",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u83B7\u53D6\u4E0A\u4E0B\u6587\u53D8\u91CF\u7684\u503C\n\u3010name\u3011\uFF1A\u4E0A\u4E0B\u6587\u53D8\u91CF\u540D\u79F0\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5BF9\u8C61(System.Object)",
        "type": "Analytical",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "getComputeJsonData",
        "name": "根据属性名获取JSON串",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u6839\u636E\u94FE\u5F0F\u5C5E\u6027\u540D\u5F97\u5230JSON\u4E32\n\u3010chainedPropName\u3011\u94FE\u5F0F\u540D\u79F0\n\u3010propertyName\u3011\u5C5E\u6027\u540D\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5BF9\u8C61(System.String)",
        "type": "Analytical",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetComputeJsonData",
        "name": "根据属性名获取JSON串",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u6839\u636E\u94FE\u5F0F\u5C5E\u6027\u540D\u5F97\u5230JSON\u4E32\n\u3010chainedPropName\u3011\u94FE\u5F0F\u540D\u79F0\n\u3010propertyName\u3011\u5C5E\u6027\u540D\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5BF9\u8C61(System.String)",
        "type": "Analytical",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetInjectedEntity",
        "name": "根据注入名获获取注入对象",
        "params": [
            {
                "key": "name",
                "name": "注入名",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u83B7\u53D6\u8FD0\u884C\u65F6\u6CE8\u5165\u7684\u5BF9\u8C61\n\u3010name\u3011\u6CE8\u5165\u540D\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5BF9\u8C61(System.Object)",
        "type": "Analytical",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetChainedPropertyValue",
        "name": "根据属性名获取属性值",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名",
                "type": "string"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u6839\u636E\u94FE\u5F0F\u5C5E\u6027\u540D\u83B7\u53D6\u5C5E\u6027\u503C\n\u3010chainedPropName\u3011\u94FE\u5F0F\u540D\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u5BF9\u8C61(System.Object)",
        "type": "Analytical",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "abs",
        "name": "绝对值",
        "params": [
            {
                "key": "value",
                "name": "数值",
                "type": "double"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u4E00\u4E2A\u6570\u503C\u7684\u7EDD\u5BF9\u503C\n\u3010value\u3011\uFF1A\u8BA1\u7B97\u6570\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Double)\n\u3010\u793A\u4F8B\u3011\uFF1Aabs(-1.5)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A1.5",
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "Round",
        "name": "四舍五入",
        "params": [
            {
                "key": "value1",
                "name": "数值",
                "type": "double"
            },
            {
                "key": "value2",
                "name": "小数位数",
                "type": "int32"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u628A\u6570\u56DB\u820D\u4E94\u5165\u4E3A\u6307\u5B9A\u7CBE\u5EA6\u7684\u6570\u503C\n\u3010value1\u3011\uFF1A\u6570\u503C \n\u3010value2\u3011\uFF1A\u56DB\u820D\u4E94\u5165\u5C0F\u6570\u4F4D\u6570\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Double)\n\u3010\u793A\u4F8B\u3011\uFF1A\nRound(1.52,1)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A1.5",
        "type": "Math",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "round",
        "name": "四舍五入",
        "params": [
            {
                "key": "value1",
                "name": "数值",
                "type": "double"
            },
            {
                "key": "value2",
                "name": "小数位数",
                "type": "int32"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u628A\u6570\u56DB\u820D\u4E94\u5165\u4E3A\u6307\u5B9A\u7CBE\u5EA6\u7684\u6570\u503C\n\u3010value1\u3011\uFF1A\u6570\u503C \n\u3010value2\u3011\uFF1A\u56DB\u820D\u4E94\u5165\u5C0F\u6570\u4F4D\u6570\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Double)\n\u3010\u793A\u4F8B\u3011\uFF1Around(1.52,1)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A1.5",
        "type": "Math",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "random",
        "name": "随机数",
        "params": [],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u968F\u673A\u751F\u6210\u5927\u4E8E\u7B49\u4E8E 0.0 \u4E14\u5C0F\u4E8E 1.0 \u7684\u4F2A\u968F\u673A\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Double)\n\u3010\u793A\u4F8B\u3011\uFF1Arandom()\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A0.624508349393",
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "sin",
        "name": "正弦",
        "params": [
            {
                "key": "value",
                "name": "数值",
                "type": "double"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u4E00\u4E2A\u6570\u5B57\u7684\u6B63\u5F26\u503C\n\u3010value\u3011\uFF1A\u8BA1\u7B97\u6570\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Double)\n\u4E00\u4E2A\u4EE5\u5F27\u5EA6\u8868\u793A\u7684\u89D2\n\u3010\u793A\u4F8B\u3011\uFF1Asin(1)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A0.84187098",
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "cos",
        "name": "余弦",
        "params": [
            {
                "key": "value",
                "name": "数值",
                "type": "double"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u4E00\u4E2A\u6570\u5B57\u7684\u4F59\u5F26\u503C\n\u3010value\u3011\uFF1A\u8BA1\u7B97\u6570\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Double)\n\u4E00\u4E2A\u4EE5\u5F27\u5EA6\u8868\u793A\u7684\u89D2\n\u3010\u793A\u4F8B\u3011\uFF1Acos(3)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A-0.9899924",
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "ceil",
        "name": "向上舍入",
        "params": [
            {
                "key": "value",
                "name": "数值",
                "type": "double"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u5927\u4E8E\u6216\u7B49\u4E8E\u51FD\u6570\u53C2\u6570\uFF0C\u5E76\u4E14\u4E0E\u4E4B\u6700\u63A5\u8FD1\u7684\u6574\u6570\n\u3010value\u3011\uFF1A\u8BA1\u7B97\u6570\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Int32)\n\u3010\u793A\u4F8B\u3011\uFF1Aceil(3.2)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A4",
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "floor",
        "name": "向下舍入",
        "params": [
            {
                "key": "value",
                "name": "数值",
                "type": "double"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u8FD4\u56DE\u5C0F\u4E8E\u6216\u7B49\u4E8E\u51FD\u6570\u53C2\u6570\uFF0C\u5E76\u4E14\u4E0E\u4E4B\u6700\u63A5\u8FD1\u7684\u6574\u6570\n\u3010value\u3011\uFF1A\u8BA1\u7B97\u6570\u503C\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Int32)\n\u3010\u793A\u4F8B\u3011\uFF1Afloor(3.2)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A3",
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "BankerRound",
        "name": "银行家舍入",
        "params": [
            {
                "key": "value1",
                "name": "数值",
                "type": "double"
            },
            {
                "key": "value2",
                "name": "小数位数",
                "type": "int32"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u4F7F\u7528\u94F6\u884C\u5BB6\u820D\u5165\u6807\u51C6\uFF0C\u628A\u6307\u5B9A\u6570\u5B57\u820D\u5165\u4E3A\u6700\u63A5\u8FD1\u7684\u6570\u5B57\n\u3010value1\u3011\uFF1A\u8981\u820D\u5165\u7684\u6570\u503C\n\u3010value2\u3011\uFF1A\u8981\u820D\u5165\u7684\u5C0F\u6570\u4F4D\u6570\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Double)\n\u3010\u793A\u4F8B\u3011\uFF1ABankerRound(36.124,2)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A36.12",
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "bankerRound",
        "name": "银行家舍入",
        "params": [
            {
                "key": "value1",
                "name": "数值",
                "type": "double"
            },
            {
                "key": "value2",
                "name": "小数位数",
                "type": "int32"
            }
        ],
        "description": "\u3010\u529F\u80FD\u3011\uFF1A\u4F7F\u7528\u94F6\u884C\u5BB6\u820D\u5165\u6807\u51C6\uFF0C\u628A\u6307\u5B9A\u6570\u5B57\u820D\u5165\u4E3A\u6700\u63A5\u8FD1\u7684\u6570\u5B57\n\u3010value1\u3011\uFF1A\u8981\u820D\u5165\u7684\u6570\u503C\n\u3010value2\u3011\uFF1A\u8981\u820D\u5165\u7684\u5C0F\u6570\u4F4D\u6570\n\u3010\u8FD4\u56DE\u503C\u7C7B\u578B\u3011\uFF1A\u6570\u503C(System.Double)\n\u3010\u793A\u4F8B\u3011\uFF1AbankerRound(36.124,2)\n\u3010\u8FD4\u56DE\u503C\u3011\uFF1A36.12",
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    }
];
