// @gsp-svc/expression/esm5/lib/schema.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * schema实体类
 * \@property entitySets 实体集集合
 * \@property entityTypes 实体类型集合
 * \@property description schema自描述
 * @author linchang 2018/12/7
 */
var /**
 * schema实体类
 * \@property entitySets 实体集集合
 * \@property entityTypes 实体类型集合
 * \@property description schema自描述
 * @author linchang 2018/12/7
 */
Schema = /** @class */ (function () {
    function Schema() {
        this.entitySets = [];
        this.entityTypes = [];
    }
    return Schema;
}());
/**
 * schema实体类
 * \@property entitySets 实体集集合
 * \@property entityTypes 实体类型集合
 * \@property description schema自描述
 * @author linchang 2018/12/7
 */
export { Schema };
if (false) {
    /** @type {?} */
    Schema.prototype.entitySets;
    /** @type {?} */
    Schema.prototype.entityTypes;
    /** @type {?} */
    Schema.prototype.description;
}
/**
 * 实体类型
 * \@property id 唯一标识
 * \@property name 语义化名称
 * \@property key 主键
 * \@property properties 属性列表信息
 * \@property navigationProperties 导航属性列表
 * \@property description 自描述
 * @author linchang 2018/12/7
 */
var /**
 * 实体类型
 * \@property id 唯一标识
 * \@property name 语义化名称
 * \@property key 主键
 * \@property properties 属性列表信息
 * \@property navigationProperties 导航属性列表
 * \@property description 自描述
 * @author linchang 2018/12/7
 */
EntityType = /** @class */ (function () {
    function EntityType() {
        this.properties = [];
        this.navigationProperties = [];
    }
    return EntityType;
}());
/**
 * 实体类型
 * \@property id 唯一标识
 * \@property name 语义化名称
 * \@property key 主键
 * \@property properties 属性列表信息
 * \@property navigationProperties 导航属性列表
 * \@property description 自描述
 * @author linchang 2018/12/7
 */
export { EntityType };
if (false) {
    /** @type {?} */
    EntityType.prototype.id;
    /** @type {?} */
    EntityType.prototype.name;
    /** @type {?} */
    EntityType.prototype.key;
    /** @type {?} */
    EntityType.prototype.properties;
    /** @type {?} */
    EntityType.prototype.navigationProperties;
    /** @type {?} */
    EntityType.prototype.description;
}
/**
 * 实体集合
 * \@property name 实体集名称
 * \@property entityType 对应的实体类型名
 * \@property description 自描述
 * @author linchang 2018/12/7
 */
var /**
 * 实体集合
 * \@property name 实体集名称
 * \@property entityType 对应的实体类型名
 * \@property description 自描述
 * @author linchang 2018/12/7
 */
EntitySet = /** @class */ (function () {
    function EntitySet() {
    }
    return EntitySet;
}());
/**
 * 实体集合
 * \@property name 实体集名称
 * \@property entityType 对应的实体类型名
 * \@property description 自描述
 * @author linchang 2018/12/7
 */
export { EntitySet };
if (false) {
    /** @type {?} */
    EntitySet.prototype.name;
    /** @type {?} */
    EntitySet.prototype.entityType;
    /** @type {?} */
    EntitySet.prototype.description;
}
/**
 * 实体类型中的属性
 * \@property name 属性名
 * \@property type 属性类型
 * \@property description 自描述
 * @author linchang 2018/12/7
 */
var /**
 * 实体类型中的属性
 * \@property name 属性名
 * \@property type 属性类型
 * \@property description 自描述
 * @author linchang 2018/12/7
 */
Property = /** @class */ (function () {
    function Property() {
    }
    return Property;
}());
/**
 * 实体类型中的属性
 * \@property name 属性名
 * \@property type 属性类型
 * \@property description 自描述
 * @author linchang 2018/12/7
 */
export { Property };
if (false) {
    /** @type {?} */
    Property.prototype.name;
    /** @type {?} */
    Property.prototype.type;
    /** @type {?} */
    Property.prototype.description;
}
/**
 * 导航属性--关联相对应实体类型或者实体集
 * \@property name 导航属性名
 * \@property type 导航属性类型
 * \@property navigationId 导航值
 * \@property description 自描述
 * @author linchang 2018/12/7
 */
var /**
 * 导航属性--关联相对应实体类型或者实体集
 * \@property name 导航属性名
 * \@property type 导航属性类型
 * \@property navigationId 导航值
 * \@property description 自描述
 * @author linchang 2018/12/7
 */
NavigationProperty = /** @class */ (function () {
    function NavigationProperty() {
    }
    return NavigationProperty;
}());
/**
 * 导航属性--关联相对应实体类型或者实体集
 * \@property name 导航属性名
 * \@property type 导航属性类型
 * \@property navigationId 导航值
 * \@property description 自描述
 * @author linchang 2018/12/7
 */
export { NavigationProperty };
if (false) {
    /** @type {?} */
    NavigationProperty.prototype.name;
    /** @type {?} */
    NavigationProperty.prototype.type;
    /** @type {?} */
    NavigationProperty.prototype.navigationId;
    /** @type {?} */
    NavigationProperty.prototype.description;
}
/** @enum {number} */
var NavigationType = {
    EntitySet: 0,
    EntityType: 1,
};
export { NavigationType };
NavigationType[NavigationType.EntitySet] = 'EntitySet';
NavigationType[NavigationType.EntityType] = 'EntityType';
