// @farris/ui-pagination/esm2015/lib/selected-list/selected-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output, Input, HostBinding } from '@angular/core';
export class PaginationSelectedListComponent {
    constructor() {
        this.data = [];
        //位置，决定是否翻转
        this.position = "bottom";
        this.close = new EventEmitter();
        this.remove = new EventEmitter();
        this.clear = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get slPosition() { return this.position == 'top'; }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClose($event) {
        $event.stopPropagation();
        this.close.emit();
    }
    /**
     * @param {?} $event
     * @param {?} id
     * @return {?}
     */
    onRemoveClick($event, id) {
        $event.stopPropagation();
        this.remove.emit(id);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClearClick($event) {
        $event.stopPropagation();
        this.clear.emit();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClick($event) {
        $event.stopPropagation();
        return false;
    }
}
PaginationSelectedListComponent.decorators = [
    { type: Component, args: [{
                selector: 'pagination-selected-list',
                template: "\r\n<div class=\"f-cmp-pager-selected-list-wrapper \">\r\n\t<div class=\"selected-list--content\" (mouseleave)=\"onClose($event)\" (click)=\"onClick($event)\">\r\n\t\t<div class=\"selected-list--arrow\"></div>\r\n\t\t<div class=\"selected-list--title\">\r\n            <div class=\"btn-group\">\r\n                <button class=\"btn btn-link\" (click)=\"onClearClick($event)\"> \r\n                    <i class=\"f-icon f-icon-remove\"></i>\r\n                    <span class=\"btn-text\">{{ 'datagrid.selectionData.clearAll' | locale: '\u6E05\u7A7A' }}</span>\r\n                </button>\r\n            </div>\r\n            <div class=\"selected-list--close\"><span class=\"f-icon f-icon-close\" (click)=\"onClose($event)\"></span></div>\r\n        </div>\r\n        <div class=\"selected-list--main\">\r\n            <ul class=\"selected-list\">  <!-- selected-list--wrap -->\r\n                <li class=\"selected-list--item ng-star-inserted\" *ngFor=\"let item of data\" style=\"min-height: 35px;\">\r\n\t\t\t\t\t<span class=\"f-icon f-icon-close-outline list-icon list-error\" (click)=\"onRemoveClick($event, item.id)\"></span>\r\n                    <div class=\"list-con\" [innerHTML]=\"item.text\"></div>\r\n                </li>\r\n                \r\n            </ul>\r\n        </div>\r\n    </div>\r\n</div>\r\n\r\n"
            }] }
];
/** @nocollapse */
PaginationSelectedListComponent.ctorParameters = () => [];
PaginationSelectedListComponent.propDecorators = {
    data: [{ type: Input }],
    position: [{ type: Input }],
    close: [{ type: Output }],
    remove: [{ type: Output }],
    clear: [{ type: Output }],
    slPosition: [{ type: HostBinding, args: ['class.pos-top',] }]
};
if (false) {
    /** @type {?} */
    PaginationSelectedListComponent.prototype.data;
    /** @type {?} */
    PaginationSelectedListComponent.prototype.position;
    /** @type {?} */
    PaginationSelectedListComponent.prototype.close;
    /** @type {?} */
    PaginationSelectedListComponent.prototype.remove;
    /** @type {?} */
    PaginationSelectedListComponent.prototype.clear;
}
