// @farris/kendo-binding/esm5/lib/directives/localization/farris_number_localization.directive.js
import { Directive, Inject, Injector } from "@angular/core";
import { UserSettingsToken } from "@farris/devkit";
import { NumberSpinnerComponent } from '@farris/ui-number-spinner';
var FarrisNumberLocalizationDirective = /** @class */ (function () {
    function FarrisNumberLocalizationDirective(injector, userSettings, componentRef) {
        this.injector = injector;
        this.userSettings = userSettings;
        this.componentRef = componentRef;
    }
    FarrisNumberLocalizationDirective.prototype.ngOnInit = function () {
        this.localize();
    };
    FarrisNumberLocalizationDirective.prototype.localize = function () {
        return;
        // 涉及金额等问题，数字格式化不予处理
        if (this.userSettings && this.userSettings.numberFormat && this.componentRef) {
            var _a = this.userSettings.numberFormat || {}, _b = _a.negativeSign, negativeSign = _b === void 0 ? null : _b, _c = _a.numberDecimalDigits, numberDecimalDigits = _c === void 0 ? null : _c, _d = _a.numberDecimalSeparator, numberDecimalSeparator = _d === void 0 ? null : _d, _e = _a.numberGroupSeparator, numberGroupSeparator = _e === void 0 ? null : _e;
            // 负号
            if (negativeSign) {
                // this.componentRef.prefix = negativeSign;
            }
            // 小数位数
            if (numberDecimalDigits !== null) {
                this.componentRef.precision = numberDecimalDigits;
                this.componentRef.formatOptions['precision'] = numberDecimalDigits;
            }
            // 小数点符号
            if (numberDecimalSeparator !== null) {
                this.componentRef.decimalSeparator = numberDecimalSeparator;
                this.componentRef.formatOptions['decimalSeparator'] = numberDecimalSeparator;
            }
            // 千分位符号
            if (numberGroupSeparator !== null) {
                this.componentRef.groupSeparator = numberGroupSeparator;
                this.componentRef.formatOptions['groupSeparator'] = numberGroupSeparator;
            }
            this.componentRef.updateValue(this.componentRef.value);
        }
    };
    FarrisNumberLocalizationDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-number-localization]'
                },] }
    ];
    /** @nocollapse */
    FarrisNumberLocalizationDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: undefined, decorators: [{ type: Inject, args: [UserSettingsToken,] }] },
        { type: NumberSpinnerComponent }
    ]; };
    return FarrisNumberLocalizationDirective;
}());
export { FarrisNumberLocalizationDirective };
