// @farris/kendo-binding/esm5/lib/directives/localization/farris_time_localization.directive.js
import { Directive, Inject, Injector } from "@angular/core";
import { UserSettingsToken } from "@farris/devkit";
import { TimePickerComponent } from '@farris/ui-time-picker';
var FarrisTimeLocalizationDirective = /** @class */ (function () {
    function FarrisTimeLocalizationDirective(injector, userSettings, componentRef) {
        this.injector = injector;
        this.userSettings = userSettings;
        this.componentRef = componentRef;
    }
    FarrisTimeLocalizationDirective.prototype.ngOnInit = function () {
        this.localize();
    };
    FarrisTimeLocalizationDirective.prototype.localize = function () {
        if (this.userSettings && this.componentRef) {
            var _a = this.userSettings.timeFormat, timeFormat = _a === void 0 ? null : _a;
            if (timeFormat) {
                this.componentRef.format = timeFormat;
            }
        }
    };
    FarrisTimeLocalizationDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-time-localization]'
                },] }
    ];
    /** @nocollapse */
    FarrisTimeLocalizationDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: undefined, decorators: [{ type: Inject, args: [UserSettingsToken,] }] },
        { type: TimePickerComponent }
    ]; };
    return FarrisTimeLocalizationDirective;
}());
export { FarrisTimeLocalizationDirective };
