// @farris/kendo-binding/esm5/lib/directives/localization/farris_date_localization.directive.js
import { Directive, Inject, Injector } from "@angular/core";
import { UserSettingsToken } from "@farris/devkit";
import { FarrisDatepickerComponent } from '@farris/ui-datepicker';
var FarrisDateLocalizationDirective = /** @class */ (function () {
    function FarrisDateLocalizationDirective(injector, userSettings, componentRef) {
        this.injector = injector;
        this.userSettings = userSettings;
        this.componentRef = componentRef;
    }
    FarrisDateLocalizationDirective.prototype.ngOnInit = function () {
        this.localize();
    };
    FarrisDateLocalizationDirective.prototype.localize = function () {
        if (this.userSettings && this.componentRef) {
            var _a = this.userSettings, _b = _a.dateFormat, dateFormat = _b === void 0 ? null : _b, _c = _a.timeFormat, timeFormat = _c === void 0 ? null : _c;
            if (dateFormat === null && timeFormat === null) {
                return;
            }
            if (timeFormat && timeFormat && this.componentRef.showTime === true) {
                this.componentRef.dateFormat = dateFormat + " " + timeFormat;
                this.componentRef.dateOpts.dateFormat = dateFormat + " " + timeFormat;
                this.componentRef.updateValue(this.componentRef.value);
            }
            else if (dateFormat) {
                this.componentRef.dateFormat = "" + dateFormat;
                this.componentRef.dateOpts.dateFormat = "" + dateFormat;
                this.componentRef.updateValue(this.componentRef.value);
            }
        }
    };
    FarrisDateLocalizationDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-date-localization]'
                },] }
    ];
    /** @nocollapse */
    FarrisDateLocalizationDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: undefined, decorators: [{ type: Inject, args: [UserSettingsToken,] }] },
        { type: FarrisDatepickerComponent }
    ]; };
    return FarrisDateLocalizationDirective;
}());
export { FarrisDateLocalizationDirective };
