// @gsp-bef/gsp-udt-metadata/esm2015/lib/UnifiedDataType/src/Element/UdtElement.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// import { BaseDataType } from '../EnumType/BaseDataType';
import { GspCommonField, GSPElementObjectType } from '../../ref';
import { GSPElementDataType } from '@gsp-bef/gsp-cm-metadata';
// import { Guid } from '../../../util/Guid';
// import { GspAssociation } from '../../ref';
// import { GSPEnumValue } from '@gsp-bef/gsp-be-metadata';
/**
 * 字段定义
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 字段定义
 */
export class UdtElement extends GspCommonField {
    constructor() {
        super();
        /**
         * 字段数据类型(默认string)
         */
        this.DataType = GSPElementDataType.String;
        // /**
        //  * 字段对象类型(默认None)
        //  */
        this.ObjectType = GSPElementObjectType.None;
        // this.Id = Guid.newGuid();
        this.Length = 36;
        this.Precision = 0;
        this.DataType = GSPElementDataType.String;
        this.ObjectType = GSPElementObjectType.None;
        this.IsMultiLanguage = false;
        this.IsRequire = false;
    }
}
if (false) {
    /**
     * 唯一标志
     * @type {?}
     */
    UdtElement.prototype.Id;
    /**
     * 编号
     * @type {?}
     */
    UdtElement.prototype.Code;
    /**
     * 名称
     * @type {?}
     */
    UdtElement.prototype.Name;
    /**
     * 默认值
     * @type {?}
     */
    UdtElement.prototype.DefaultValue;
    /**
     * 数据标识
     * @type {?}
     */
    UdtElement.prototype.LabelId;
    /**
     * 关联字段Id
     * @type {?}
     */
    UdtElement.prototype.RefElementId;
    /**
     * 是否关联带出字段
     * @type {?}
     */
    UdtElement.prototype.IsRefElement;
    /**
     * 字段数据类型(默认string)
     * @type {?}
     */
    UdtElement.prototype.DataType;
    /**
     * 长度
     * @type {?}
     */
    UdtElement.prototype.Length;
    /**
     * 精度
     * @type {?}
     */
    UdtElement.prototype.Precision;
    /**
     * 是否唯一
     * @type {?}
     */
    UdtElement.prototype.IsUnique;
    /**
     * 是否必须
     * @type {?}
     */
    UdtElement.prototype.IsRequire;
    /**
     * 是否启用多语
     * @type {?}
     */
    UdtElement.prototype.IsMultiLanguage;
    /** @type {?} */
    UdtElement.prototype.ObjectType;
    /**
     * 包含的关联列表
     * @type {?}
     */
    UdtElement.prototype.ChildAssociations;
    /**
     * 包含的枚举列表
     * @type {?}
     */
    UdtElement.prototype.ContainEnumValues;
    /**
     * 是否启用业务字段
     * @type {?}
     */
    UdtElement.prototype.IsUdt;
    /**
     * 业务字段包名
     * @type {?}
     */
    UdtElement.prototype.UdtPkgName;
    /**
     * 业务字段ID
     * @type {?}
     */
    UdtElement.prototype.UdtID;
    /**
     * 引用的业务字段，前端使用(仅记录单值)
     * @type {?}
     */
    UdtElement.prototype.UnifiedDataType;
    /**
     * udt名称
     * @type {?}
     */
    UdtElement.prototype.UdtName;
    /**
     * 多数据库列
     * @type {?}
     */
    UdtElement.prototype.ChildElements;
    /**
     * 多数据库列映射关系
     * @type {?}
     */
    UdtElement.prototype.MappingRelation;
    /**
     * 是否udt关联带出字段
     * @type {?}
     */
    UdtElement.prototype.IsFromAssoUdt;
}
