// @farris/ui-number-spinner/bundles/farris-ui-number-spinner.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@farris/ui-common'), require('@angular/common'), require('@farris/ui-popover'), require('bignumber.js'), require('@angular/core'), require('@angular/forms'), require('@farris/ui-locale')) :
    typeof define === 'function' && define.amd ? define('@farris/ui-number-spinner', ['exports', '@farris/ui-common', '@angular/common', '@farris/ui-popover', 'bignumber.js', '@angular/core', '@angular/forms', '@farris/ui-locale'], factory) :
    (factory((global.farris = global.farris || {}, global.farris['ui-number-spinner'] = {}),global.uiCommon,global.ng.common,global.uiPopover,global.bignumber_js,global.ng.core,global.ng.forms,global.uiLocale));
}(this, (function (exports,uiCommon,common,uiPopover,bignumber_js,i0,forms,uiLocale) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NumberSpinnerService = /** @class */ (function () {
        function NumberSpinnerService() {
        }
        NumberSpinnerService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */
        NumberSpinnerService.ctorParameters = function () { return []; };
        /** @nocollapse */ NumberSpinnerService.ngInjectableDef = i0.defineInjectable({ factory: function NumberSpinnerService_Factory() { return new NumberSpinnerService(); }, token: NumberSpinnerService, providedIn: "root" });
        return NumberSpinnerService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NumberSpinnerComponent = /** @class */ (function () {
        function NumberSpinnerComponent(el, render, injector) {
            this.el = el;
            this.render = render;
            this.injector = injector;
            this.hostClass = true;
            this.disabled = false;
            this.readonly = false;
            this.editable = true;
            // 空白提示文本
            this.placeholder = '';
            // up or down 步长
            this.step = 1;
            /**
             * 启用大数支持
             */
            this.bigNumber = false;
            // 是否显示加减按钮
            this.showButton = true;
            // 是否使用千分值
            this.useThousands = true;
            // 文本方向
            this.textAlign = 'left';
            // 自动补全小数
            this.autoDecimal = true;
            // 允许为空
            this.canNull = false;
            // 精度
            this.precision = 0;
            // 前缀
            this.prefix = '';
            // 后缀
            this.suffix = '';
            /**
             * 小数点符号
             */
            this.decimalSeparator = '.';
            /**
             * 千分位符号
             */
            this.groupSeparator = ',';
            /**
             * 使用千分位时，每组显示的字符数
             */
            this.groupSize = 3;
            this._value = '';
            /**
             * 显示0值
             */
            this.showZero = true;
            this.valueChange = new i0.EventEmitter();
            this.blur = new i0.EventEmitter();
            this.focus = new i0.EventEmitter();
            this.isFocus = false;
            this.popValue = '';
            this.isActiveTip = false;
            this.formatOptions = {};
            this._realValue = null;
            this.showtipTimer = null;
            this._updateOn = 'change';
            this.localeService = null;
            this.onTouchedCallback = ( /**
             * @return {?}
             */function () { });
            this.onChangeCallback = ( /**
             * @return {?}
             */function () { });
            this.cdRef = this.injector.get(i0.ChangeDetectorRef, null);
            this.localeService = this.injector.get(uiLocale.LocaleService, null);
            this.commonUtils = this.injector.get(uiCommon.CommonUtils, new uiCommon.CommonUtils());
        }
        Object.defineProperty(NumberSpinnerComponent.prototype, "value", {
            get: /**
             * @return {?}
             */ function () {
                return this._value;
            },
            set: /**
             * @param {?} val
             * @return {?}
             */ function (val) {
                this._value = val;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        NumberSpinnerComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.formatOptions = this.buildFormatOptions();
                this._ngControl = this.injector.get(forms.NgControl, null);
                if (this.localeService) {
                    /** @type {?} */
                    var defaultPlaceHolder = this.localeService.getValue('numberSpinner.placeholder') || '请输入数字';
                    if (!this.placeholder) {
                        this.placeholder = defaultPlaceHolder;
                    }
                }
            };
        /**
         * @return {?}
         */
        NumberSpinnerComponent.prototype.ngAfterViewInit = /**
         * @return {?}
         */
            function () {
                if (this._ngControl && this._ngControl.control) {
                    this._updateOn = this._ngControl.control.updateOn;
                }
                this.listenInputPasteEvent();
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        NumberSpinnerComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                if (changes.value && !changes.value.isFirstChange()) {
                    this._realValue = this.getRealValue(changes.value.currentValue);
                    this.value = this.format(this._realValue);
                }
                if (changes.showZero && !changes.showZero.isFirstChange()) {
                    this.value = this.format(this._realValue);
                }
                if (changes.precision && !changes.precision.isFirstChange()) {
                    this.onOptionsChanged();
                }
                if (changes.useThousands && !changes.useThousands.isFirstChange()) {
                    this.onOptionsChanged();
                }
                if (changes.prefix && !changes.prefix.isFirstChange()) {
                    this.onOptionsChanged();
                }
                if (changes.suffix && !changes.suffix.isFirstChange()) {
                    this.onOptionsChanged();
                }
            };
        /**
         * @private
         * @return {?}
         */
        NumberSpinnerComponent.prototype.onOptionsChanged = /**
         * @private
         * @return {?}
         */
            function () {
                this.formatOptions = this.buildFormatOptions();
                this.value = this.format(this._realValue);
            };
        /**
         * @return {?}
         */
        NumberSpinnerComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () { };
        // 支持粘贴带格式的数据
        // 支持粘贴带格式的数据
        /**
         * @return {?}
         */
        NumberSpinnerComponent.prototype.listenInputPasteEvent =
            // 支持粘贴带格式的数据
            /**
             * @return {?}
             */
            function () {
                var _this = this;
                this.input.nativeElement.addEventListener('paste', ( /**
                 * @param {?} event
                 * @return {?}
                 */function (event) {
                    event.preventDefault();
                    /** @type {?} */
                    var clipboardData = event.clipboardData || window['clipboardData'];
                    /** @type {?} */
                    var pasteValue = clipboardData.getData('text');
                    /** @type {?} */
                    var val = _this.cleanNumString(pasteValue);
                    if (_this.isEmpty(val)) {
                        return;
                    }
                    /** @type {?} */
                    var target = event.target;
                    /** @type {?} */
                    var start = (( /** @type {?} */(target))).selectionStart;
                    /** @type {?} */
                    var end = (( /** @type {?} */(target))).selectionEnd;
                    if (_this.isEmpty(val)) {
                        target.value = '';
                    }
                    else {
                        target.value = target.value.slice(0, start) + val + target.value.slice(end);
                    }
                    if (_this._updateOn === 'change') {
                        _this._realValue = _this.getRealValue(target.value);
                        _this.value = _this._realValue;
                        _this._modelChanged(_this._realValue);
                    }
                }));
            };
        /**
         * @param {?} $event
         * @param {?=} type
         * @return {?}
         */
        NumberSpinnerComponent.prototype.onBlur = /**
         * @param {?} $event
         * @param {?=} type
         * @return {?}
         */
            function ($event, type) {
                if (type === void 0) {
                    type = '';
                }
                if (this.readonly || this.disabled) {
                    return;
                }
                if (this._updateOn === 'blur') {
                    /** @type {?} */
                    var val = this.cleanNumString($event.value);
                    this._realValue = this.getRealValue(val);
                }
                this.value = this.format(this._realValue);
                this.input.nativeElement.value = this.value;
                this.isFocus = false;
                this.onModelChange(this._realValue, 'blur');
                this.blur.emit({ event: $event, formatted: this.value, value: this._realValue, instance: this });
            };
        /**
         * @param {?} $event
         * @return {?}
         */
        NumberSpinnerComponent.prototype.onClick = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                $event.stopPropagation();
            };
        /**
         * @param {?} $event
         * @return {?}
         */
        NumberSpinnerComponent.prototype.onFocus = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                if (this.readonly || this.disabled) {
                    this.isFocus = false;
                    return;
                }
                this.value = this.isEmpty(this._realValue) ? '' : ((!this.showZero && this._realValue == '0') ? '' : this._realValue);
                this.isFocus = true;
                this.focus.emit({ event: $event, formatted: this.value, value: this._realValue, instance: this });
            };
        /**
         * @param {?} $event
         * @return {?}
         */
        NumberSpinnerComponent.prototype.onMouseEnter = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                if (this.value) {
                    this.popValue = this.value;
                    this.isActiveTip = this.isShowPopover();
                }
            };
        /**
         * @param {?} e
         * @return {?}
         */
        NumberSpinnerComponent.prototype.onKeyDown = /**
         * @param {?} e
         * @return {?}
         */
            function (e) {
                if (e.key === 'ArrowDown') {
                    e.preventDefault();
                    this.down(e);
                    e.stopPropagation();
                }
                if (e.key === 'ArrowUp') {
                    e.preventDefault();
                    this.up(e);
                    e.stopPropagation();
                }
            };
        /**
         * @private
         * @param {?} realVal
         * @return {?}
         */
        NumberSpinnerComponent.prototype._modelChanged = /**
         * @private
         * @param {?} realVal
         * @return {?}
         */
            function (realVal) {
                this._realValue = realVal;
                this.onChangeCallback(realVal);
                this.onTouchedCallback();
                this.valueChange.emit(realVal);
            };
        /**
         * @param {?} realVal
         * @param {?=} updateOn
         * @return {?}
         */
        NumberSpinnerComponent.prototype.onModelChange = /**
         * @param {?} realVal
         * @param {?=} updateOn
         * @return {?}
         */
            function (realVal, updateOn) {
                if (updateOn === void 0) {
                    updateOn = 'change';
                }
                // this._realValue = this.getRealValue(val);
                /** @type {?} */
                var _realValue = realVal;
                if (updateOn === 'change') {
                    /** @type {?} */
                    var num = this.cleanNumString('' + realVal);
                    _realValue = this.getRealValue(num);
                }
                this.isActiveTip = false;
                this.popValue = this.format(_realValue);
                if (this._updateOn === updateOn) {
                    this._modelChanged(_realValue);
                }
            };
        /**
         * @param {?} e
         * @param {?=} type
         * @return {?}
         */
        NumberSpinnerComponent.prototype.up = /**
         * @param {?} e
         * @param {?=} type
         * @return {?}
         */
            function (e, type) {
                if (type === void 0) {
                    type = null;
                }
                this.compute('up');
                e.stopPropagation();
            };
        /**
         * @param {?} e
         * @param {?=} type
         * @return {?}
         */
        NumberSpinnerComponent.prototype.down = /**
         * @param {?} e
         * @param {?=} type
         * @return {?}
         */
            function (e, type) {
                if (type === void 0) {
                    type = null;
                }
                this.compute('down');
                e.stopPropagation();
            };
        /**
         * @param {?=} tye
         * @return {?}
         */
        NumberSpinnerComponent.prototype.compute = /**
         * @param {?=} tye
         * @return {?}
         */
            function (tye) {
                if (tye === void 0) {
                    tye = 'up';
                }
                if (this.isDisableOfBtn(tye)) {
                    /** @type {?} */
                    var _resultValue = void 0;
                    /** @type {?} */
                    var realBigNum = new bignumber_js.BigNumber(this._realValue || 0);
                    if (tye === 'up') {
                        _resultValue = realBigNum.plus(Number(this.step));
                    }
                    else {
                        _resultValue = realBigNum.minus(Number(this.step));
                    }
                    /** @type {?} */
                    var s = _resultValue.toFixed();
                    if (!this.isFocus) {
                        this.value = this.format(s);
                    }
                    else {
                        this.value = s;
                    }
                    this.input.nativeElement.value = this.value;
                    this._modelChanged(this.getRealValue(_resultValue));
                }
            };
        /**
         * @param {?} type
         * @param {?=} value
         * @return {?}
         */
        NumberSpinnerComponent.prototype.isDisableOfBtn = /**
         * @param {?} type
         * @param {?=} value
         * @return {?}
         */
            function (type, value) {
                if (value === undefined) {
                    value = this._realValue;
                }
                value = new bignumber_js.BigNumber(value);
                if (type === 'up' && !(new bignumber_js.BigNumber(this.max)).isNaN() && value.gte(this.max)) {
                    return false;
                }
                if (type === 'down' && !(new bignumber_js.BigNumber(this.min)).isNaN() && value.lte(this.min)) {
                    return false;
                }
                return true;
            };
        /**
         * @return {?}
         */
        NumberSpinnerComponent.prototype.isShowPopover = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var width = this.input.nativeElement.clientWidth;
                var _a = window.getComputedStyle(this.input.nativeElement), paddingLeft = _a.paddingLeft, paddingRight = _a.paddingRight;
                /** @type {?} */
                var _width = width - (Number(paddingLeft.replace('px', '')) || 0) - (Number(paddingRight.replace('px', '')) || 0);
                /** @type {?} */
                var txtWidth = this.commonUtils.getTextWidth(this.value, window.getComputedStyle(this.input.nativeElement).font);
                if (_width && _width < txtWidth) {
                    return true;
                }
                return false;
            };
        /**
         * @return {?}
         */
        NumberSpinnerComponent.prototype._getPrecision = /**
         * @return {?}
         */
            function () {
                return Number(this.precision || 0);
            };
        /**
         * @param {?} n
         * @return {?}
         */
        NumberSpinnerComponent.prototype.toFixed = /**
         * @param {?} n
         * @return {?}
         */
            function (n) {
                /** @type {?} */
                var _n = n;
                if (!bignumber_js.BigNumber.isBigNumber(n)) {
                    _n = new bignumber_js.BigNumber(n);
                }
                if (this.precision !== null && this.precision !== undefined) {
                    return n.toFixed(this._getPrecision());
                }
                return n.toFixed();
            };
        /**
         * @param {?} _n
         * @return {?}
         */
        NumberSpinnerComponent.prototype._getRealValue = /**
         * @param {?} _n
         * @return {?}
         */
            function (_n) {
                /** @type {?} */
                var n = this.toFixed(_n);
                return this.bigNumber ? n : Number(n);
            };
        /**
         * @param {?} val
         * @return {?}
         */
        NumberSpinnerComponent.prototype.getRealValue = /**
         * @param {?} val
         * @return {?}
         */
            function (val) {
                if (this.parser) {
                    if (!isNaN(Number(val))) {
                        return val;
                    }
                    else {
                        return this.parser(val);
                    }
                }
                /** @type {?} */
                var _n = this.validInterval(new bignumber_js.BigNumber(val));
                if (_n.isNaN()) {
                    if (this.canNull) {
                        return null;
                    }
                    else {
                        /** @type {?} */
                        var minBigNum = new bignumber_js.BigNumber('' + this.min);
                        /** @type {?} */
                        var maxBigNum = new bignumber_js.BigNumber('' + this.max);
                        if (!minBigNum.isNaN()) {
                            _n = minBigNum;
                        }
                        else if (!maxBigNum.isNaN()) {
                            _n = maxBigNum;
                        }
                        else {
                            return 0;
                        }
                    }
                    // if (this.canNull || minBigNum.isNaN()) {
                    //     return null;
                    // } else {
                    //     _n = minBigNum;
                    // }
                }
                return this._getRealValue(_n);
            };
        /**
         * @private
         * @return {?}
         */
        NumberSpinnerComponent.prototype.buildFormatOptions = /**
         * @private
         * @return {?}
         */
            function () {
                return {
                    prefix: this.prefix,
                    suffix: this.suffix,
                    decimalSeparator: this.decimalSeparator,
                    groupSeparator: this.useThousands ? this.groupSeparator : '',
                    groupSize: this.groupSize
                };
            };
        /**
         * @param {?} val
         * @return {?}
         */
        NumberSpinnerComponent.prototype.isEmpty = /**
         * @param {?} val
         * @return {?}
         */
            function (val) {
                return isNaN(val) || val === null || val === undefined || val === '';
            };
        /**
         * @param {?} bn
         * @param {?=} emit
         * @return {?}
         */
        NumberSpinnerComponent.prototype.validInterval = /**
         * @param {?} bn
         * @param {?=} emit
         * @return {?}
         */
            function (bn, emit) {
                if (emit === void 0) {
                    emit = true;
                }
                /** @type {?} */
                var _bnVal = bn;
                if (!this.isEmpty(this.max)) {
                    /** @type {?} */
                    var _maxBigNum = new bignumber_js.BigNumber('' + this.max);
                    if (bn.gt(_maxBigNum)) {
                        _bnVal = _maxBigNum;
                        /** @type {?} */
                        var _realValue = this._getRealValue(_maxBigNum);
                        if (emit) {
                            this._modelChanged(_realValue);
                        }
                    }
                }
                if (!this.isEmpty(this.min)) {
                    /** @type {?} */
                    var _minBigNum = new bignumber_js.BigNumber('' + this.min);
                    if (bn.lt(_minBigNum)) {
                        _bnVal = _minBigNum;
                        /** @type {?} */
                        var _realValue = this._getRealValue(_minBigNum);
                        if (emit) {
                            this._modelChanged(_realValue);
                        }
                    }
                }
                return _bnVal;
            };
        /**
         * @param {?} val
         * @param {?=} emit
         * @return {?}
         */
        NumberSpinnerComponent.prototype.format = /**
         * @param {?} val
         * @param {?=} emit
         * @return {?}
         */
            function (val, emit) {
                if (emit === void 0) {
                    emit = true;
                }
                val = this.cleanNumString(val);
                /** @type {?} */
                var bigVal = new bignumber_js.BigNumber(val);
                /** @type {?} */
                var _bgNum = bigVal;
                if (emit) {
                    _bgNum = this.validInterval(bigVal);
                }
                if (_bgNum.valueOf() == '0' && !this.showZero) {
                    return '';
                }
                if (this.canNull && bigVal.isNaN()) {
                    return '';
                }
                else {
                    if (_bgNum.isNaN()) {
                        return '';
                    }
                }
                if (this.formatter) {
                    return this.formatter(_bgNum.toNumber());
                }
                else {
                    if (!Object.keys(this.formatOptions).length) {
                        this.formatOptions = this.buildFormatOptions();
                    }
                    return this._toFormat(_bgNum, this.formatOptions);
                }
            };
        /**
         * @param {?} _bgNum
         * @param {?} fmt
         * @return {?}
         */
        NumberSpinnerComponent.prototype._toFormat = /**
         * @param {?} _bgNum
         * @param {?} fmt
         * @return {?}
         */
            function (_bgNum, fmt) {
                if (this.precision !== null && this.precision !== undefined) {
                    return _bgNum.toFormat(this._getPrecision(), fmt);
                }
                else {
                    return _bgNum.toFormat(fmt);
                }
            };
        /**
         * @param {?} val
         * @return {?}
         */
        NumberSpinnerComponent.prototype.cleanNumString = /**
         * @param {?} val
         * @return {?}
         */
            function (val) {
                val = (val === null || val === undefined || val === '') ? '' : String(val);
                val = val.replace(new RegExp(this.prefix, 'g'), '')
                    .replace(new RegExp(this.suffix, 'g'), '').replace(/\,/g, '');
                if (this.groupSeparator && this.groupSeparator !== ',') {
                    val = val.replace(new RegExp("\\" + this.groupSeparator, 'g'), '');
                }
                if (this.decimalSeparator && this.decimalSeparator !== '.') {
                    val = val.replace(new RegExp("\\" + this.decimalSeparator, 'g'), '.');
                }
                return val;
            };
        /**
         * @param {?} val
         * @return {?}
         */
        NumberSpinnerComponent.prototype.updateValue = /**
         * @param {?} val
         * @return {?}
         */
            function (val) {
                val = this.cleanNumString(val);
                this._realValue = this.getRealValue(val);
                this.value = this.format(this._realValue);
                this.el.nativeElement.value = this.value;
            };
        /**
         * @private
         * @return {?}
         */
        NumberSpinnerComponent.prototype.updateControlValue = /**
         * @private
         * @return {?}
         */
            function () {
                /** @type {?} */
                var _val = this.cleanNumString(this.value);
                if (this.parser) {
                    _val = this.parser(this.value);
                }
                if (this.isEmpty(this._realValue) && this.isEmpty(_val)) {
                    return;
                }
                /** @type {?} */
                var rv = new bignumber_js.BigNumber(this._realValue);
                /** @type {?} */
                var cv = new bignumber_js.BigNumber(_val);
                if (rv.isNaN() && cv.isNaN()) {
                    return;
                }
                // if (!rv.eq(cv)) {
                // }
                if (this._ngControl) {
                    /** @type {?} */
                    var formgroup = this._ngControl['formDirective'];
                    if (formgroup) {
                        /** @type {?} */
                        var ctrl = formgroup.control.get(this._ngControl.name);
                        if (!ctrl && formgroup.control.controls) {
                            ctrl = formgroup.control.controls[this._ngControl.name];
                        }
                        if (ctrl) {
                            if (ctrl.value !== this._realValue) {
                                ctrl.setValue(this._realValue);
                            }
                        }
                    }
                }
            };
        /**
         * @param {?} val
         * @return {?}
         */
        NumberSpinnerComponent.prototype.writeValue = /**
         * @param {?} val
         * @return {?}
         */
            function (val) {
                this._realValue = val;
                this.value = this.format(this._realValue, false);
                this.updateControlValue();
            };
        /**
         * @param {?} fn
         * @return {?}
         */
        NumberSpinnerComponent.prototype.registerOnChange = /**
         * @param {?} fn
         * @return {?}
         */
            function (fn) {
                this.onChangeCallback = fn;
            };
        /**
         * @param {?} fn
         * @return {?}
         */
        NumberSpinnerComponent.prototype.registerOnTouched = /**
         * @param {?} fn
         * @return {?}
         */
            function (fn) {
                this.onTouchedCallback = fn;
            };
        /**
         * @param {?} isDisabled
         * @return {?}
         */
        NumberSpinnerComponent.prototype.setDisabledState = /**
         * @param {?} isDisabled
         * @return {?}
         */
            function (isDisabled) {
                this.disabled = isDisabled;
            };
        NumberSpinnerComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'farris-number-spinner',
                        template: "\n        <div class=\"input-group  f-state-hover flex-row\"\n            [ngClass]=\"{'f-state-disabled': disabled,'f-state-readonly': readonly,'f-state-focus': isFocus }\"\n            (mouseenter)=\"onMouseEnter($event)\"\n            #popover=\"bs-popover\"\n            farrisPopover\n            [popover]=\"popValue\" [triggers]=\"'hover'\" [popActive]=\"isActiveTip\"\n            [showAction]=\"'show2'\"  [container]=\"'body'\">\n            <input #input\n                type=\"text\"\n                number-only\n                [precision]=\"precision\"\n                autocomplete=\"off\"\n                class=\"form-control f-utils-fill\"\n                [ngModel]=\"value\"\n                [placeholder]=\"disabled||readonly || !editable?'':placeholder\"\n                [readonly]=\"readonly || !editable\"\n                [disabled]=\"disabled\"\n                [style.padding-right]=\"'4px'\"\n                [style.text-align]=\"textAlign\"\n                (ngModelChange)=\"onModelChange($event)\"\n                (blur)=\"onBlur(input)\"\n                (focus)=\"onFocus($event)\"\n                (keydown)=\"onKeyDown($event)\"\n                (click)=\"onClick($event)\"\n            />\n            <div class=\"input-group-append btn-group btn-group-number m-0\" *ngIf=\"!disabled && !readonly && showButton\">\n                <span class=\"icon k-i-arrow-chevron-up number-arrow-chevron\" [style.cursor]=\"isDisableOfBtn('up') ? 'pointer' : 'not-allowed'\"\n                        (click)=\"up($event)\"></span>\n                <span class=\"icon k-i-arrow-chevron-down number-arrow-chevron\" [style.cursor]=\"isDisableOfBtn('down') ? 'pointer' : 'not-allowed'\"\n                        (click)=\"down($event)\"></span>\n                \n            </div>\n        </div>\n    ",
                        providers: [
                            {
                                provide: forms.NG_VALUE_ACCESSOR,
                                useExisting: i0.forwardRef(( /**
                                 * @return {?}
                                 */function () { return NumberSpinnerComponent; })),
                                multi: true
                            }
                        ],
                        styles: [".input-group .btn-group-number{height:1.50003rem;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;background-color:#fff;border-left:1px solid #d9d9d9;width:1.50003rem}.input-group .btn-group-number .number-arrow-chevron{-webkit-box-flex:1;flex:1;line-height:1;height:50%;text-align:center}.input-group .btn-group-number .number-arrow-chevron:active{background-color:#e6e9f0}.input-group .btn-group-number .number-arrow-chevron:nth-child(2){border-top:1px solid #d9d9d9}input::-webkit-inner-spin-button,input::-webkit-outer-spin-button{-webkit-appearance:none!important;margin:0}.number-range{position:relative}.number-range .input-container{display:-webkit-box;display:flex;padding:0}.number-range .input-container .sub-input-group{-webkit-box-flex:1;flex:1;position:relative;display:-webkit-box;display:flex;-webkit-transition:.3s ease-out;transition:.3s ease-out}.number-range .input-container .sub-input-group .sub-input{width:100%;border:none;outline:0;background-color:rgba(0,0,0,0);min-width:2px;padding:.125rem 4px .125rem .5rem}.number-range .input-container .spliter{width:15px;text-align:center}"]
                    }] }
        ];
        /** @nocollapse */
        NumberSpinnerComponent.ctorParameters = function () {
            return [
                { type: i0.ElementRef },
                { type: i0.Renderer2 },
                { type: i0.Injector }
            ];
        };
        NumberSpinnerComponent.propDecorators = {
            hostClass: [{ type: i0.HostBinding, args: ['class.f-cmp-number-spinner',] }],
            disabled: [{ type: i0.Input }],
            readonly: [{ type: i0.Input }],
            editable: [{ type: i0.Input }],
            formatter: [{ type: i0.Input }],
            parser: [{ type: i0.Input }],
            placeholder: [{ type: i0.Input }],
            step: [{ type: i0.Input }],
            max: [{ type: i0.Input }],
            min: [{ type: i0.Input }],
            bigNumber: [{ type: i0.Input }],
            showButton: [{ type: i0.Input }],
            useThousands: [{ type: i0.Input }],
            textAlign: [{ type: i0.Input }],
            autoDecimal: [{ type: i0.Input }],
            canNull: [{ type: i0.Input }],
            precision: [{ type: i0.Input }],
            prefix: [{ type: i0.Input }],
            suffix: [{ type: i0.Input }],
            decimalSeparator: [{ type: i0.Input }],
            groupSeparator: [{ type: i0.Input }],
            groupSize: [{ type: i0.Input }],
            value: [{ type: i0.Input }],
            showZero: [{ type: i0.Input }],
            valueChange: [{ type: i0.Output }],
            blur: [{ type: i0.Output }],
            focus: [{ type: i0.Output }],
            input: [{ type: i0.ViewChild, args: ['input',] }]
        };
        return NumberSpinnerComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NumberOnlyDirective = /** @class */ (function () {
        function NumberOnlyDirective(element) {
            this.element = element;
        }
        /**
         * @param {?} e
         * @return {?}
         */
        NumberOnlyDirective.prototype.onKeyDown = /**
         * @param {?} e
         * @return {?}
         */
            function (e) {
                /** @type {?} */
                var val = (( /** @type {?} */(e.target))).value;
                /** @type {?} */
                var key = e.key;
                /** @type {?} */
                var checkKeys = ['.', '-', '%'];
                /** @type {?} */
                var selectionVal = this.getSelectionValue(e.target);
                if (val && checkKeys.indexOf(key) > -1 && val.indexOf(key) > -1 && !(selectionVal && selectionVal.indexOf(key) > -1)) {
                    e.preventDefault();
                }
                if ([46, 8, 9, 27, 13, 110, 190, 173].indexOf(e.keyCode) !== -1 ||
                    // 允许全选: Ctrl+A
                    (e.keyCode === 65 && (e.ctrlKey || e.metaKey)) ||
                    // 允许复制: Ctrl+C
                    (e.keyCode === 67 && (e.ctrlKey || e.metaKey)) ||
                    // 允许粘贴: Ctrl+V
                    (e.keyCode === 86 && (e.ctrlKey || e.metaKey)) ||
                    // 允许剪切: Ctrl+X
                    (e.keyCode === 88 && (e.ctrlKey || e.metaKey)) ||
                    // 允许: home（头）, end（尾）, left（左移）, right（右移）
                    (e.keyCode >= 35 && e.keyCode <= 39) ||
                    e.keyCode === 189 || e.keyCode === 109) {
                    // let it happen, don't do anything
                    return;
                }
                // 确保数字以外的案件被拒绝执行默认动作
                if ((e.shiftKey || (e.keyCode < 48 || e.keyCode > 57)) && (e.keyCode < 96 || e.keyCode > 105)) {
                    e.preventDefault();
                }
                // 判断当前小数精度
                if (val.indexOf('.') > -1 && this.precision) {
                    /** @type {?} */
                    var curPos = (( /** @type {?} */(e.target))).selectionStart;
                    /** @type {?} */
                    var selectionEnd = (( /** @type {?} */(e.target))).selectionEnd;
                    /** @type {?} */
                    var _precision = val.split('.')[1].length;
                    /** @type {?} */
                    var dotPos = val.indexOf('.');
                    if (curPos > dotPos) {
                        /** @type {?} */
                        var selectionText = val.substring(curPos, selectionEnd);
                        if (_precision - selectionText.length >= Number(this.precision)) {
                            e.preventDefault();
                            return false;
                        }
                    }
                }
            };
        /**
         * @private
         * @param {?} target
         * @return {?}
         */
        NumberOnlyDirective.prototype.getSelectionValue = /**
         * @private
         * @param {?} target
         * @return {?}
         */
            function (target) {
                /** @type {?} */
                var start = (( /** @type {?} */(target))).selectionStart;
                /** @type {?} */
                var end = (( /** @type {?} */(target))).selectionEnd;
                return target.value.substring(start, end);
            };
        NumberOnlyDirective.decorators = [
            { type: i0.Directive, args: [{
                        // tslint:disable-next-line: directive-selector
                        selector: '[number-only]'
                    },] }
        ];
        /** @nocollapse */
        NumberOnlyDirective.ctorParameters = function () {
            return [
                { type: i0.ElementRef }
            ];
        };
        NumberOnlyDirective.propDecorators = {
            precision: [{ type: i0.Input }],
            onKeyDown: [{ type: i0.HostListener, args: ['keydown', ['$event'],] }]
        };
        return NumberOnlyDirective;
    }());

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NumberRangeComponent = /** @class */ (function (_super) {
        __extends(NumberRangeComponent, _super);
        function NumberRangeComponent() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.beginPlaceHolder = '';
            _this.endPlaceHolder = '';
            _this.separator = '~';
            _this.__value = '';
            _this.beginValueChange = new i0.EventEmitter();
            _this.endValueChange = new i0.EventEmitter();
            /* 开始数字展示值 */
            _this.prevValue = '';
            /* 结束数字展示值 */
            _this.nextValue = '';
            _this._beginValue = '';
            _this._endValue = '';
            _this.focusState = {
                isFocus: false,
                target: '' // begin | end
            };
            return _this;
            //#endregion
        }
        Object.defineProperty(NumberRangeComponent.prototype, "beginValue", {
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this.prevValue = this.format(value);
                // 开始真实数值
                this._beginValue = this.isEmpty(value) ? '' : value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NumberRangeComponent.prototype, "endValue", {
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this.nextValue = this.format(value);
                // 结束真实数值
                this._endValue = this.isEmpty(value) ? '' : value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NumberRangeComponent.prototype, "value", {
            get: /**
             * @return {?}
             */ function () {
                return this.__value;
            },
            set: /**
             * @param {?} v
             * @return {?}
             */ function (v) {
                if (v !== this.__value) {
                    this.__value = v === '~' ? '' : v;
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NumberRangeComponent.prototype, "preNumValue", {
            get: /**
             * @return {?}
             */ function () {
                /** @type {?} */
                var v = this.getRealValue(this.parser ? this.prevValue : this._beginValue);
                if (this.canNull && this.isEmpty(v)) {
                    return '';
                }
                else {
                    return v || 0;
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NumberRangeComponent.prototype, "nextNumValue", {
            get: /**
             * @return {?}
             */ function () {
                /** @type {?} */
                var v = this.getRealValue(this.parser ? this.nextValue : this._endValue);
                if (this.canNull && this.isEmpty(v)) {
                    return '';
                }
                else {
                    return v || 0;
                }
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        NumberRangeComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                _super.prototype.ngOnInit.call(this);
                this.showButton = false;
                this.initPlaceHolderInfo();
            };
        /**
         * @return {?}
         */
        NumberRangeComponent.prototype.ngAfterViewInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.documentClick = this.render.listen(document, 'click', ( /**
                 * @param {?} e
                 * @return {?}
                 */function (e) {
                    _this.render.setStyle(_this.prevWrapper.nativeElement, 'flex', 1);
                    _this.render.setStyle(_this.nextWrapper.nativeElement, 'flex', 1);
                }));
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        NumberRangeComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                _super.prototype.ngOnChanges.call(this, changes);
                if (changes.value && !changes.value.isFirstChange()) {
                    this.setValue(changes.value.currentValue);
                }
            };
        /**
         * @return {?}
         */
        NumberRangeComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                if (this.documentClick) {
                    this.documentClick();
                }
            };
        /**
         * @private
         * @return {?}
         */
        NumberRangeComponent.prototype.initPlaceHolderInfo = /**
         * @private
         * @return {?}
         */
            function () {
                if (!this.beginPlaceHolder) {
                    this.beginPlaceHolder = this.localeService.getValue('numberSpinner.range.begin') || this.placeholder;
                }
                if (!this.endPlaceHolder) {
                    this.endPlaceHolder = this.localeService.getValue('numberSpinner.range.end') || this.placeholder;
                }
            };
        /**
         * @private
         * @return {?}
         */
        NumberRangeComponent.prototype.getRangeValue = /**
         * @private
         * @return {?}
         */
            function () {
                return this.preNumValue + this.separator + this.nextNumValue;
            };
        /**
         * @private
         * @return {?}
         */
        NumberRangeComponent.prototype.getEndValue = /**
         * @private
         * @return {?}
         */
            function () {
                return this.getRealValue(this.nextValue);
            };
        /**
         * @private
         * @return {?}
         */
        NumberRangeComponent.prototype.getBeginValue = /**
         * @private
         * @return {?}
         */
            function () {
                return this.getRealValue(this.prevValue);
            };
        /**
         * @param {?} val
         * @return {?}
         */
        NumberRangeComponent.prototype.getRealValue = /**
         * @param {?} val
         * @return {?}
         */
            function (val) {
                if (this.parser) {
                    return this.parser(val);
                }
                /** @type {?} */
                var _n = this.validInterval(new bignumber_js.BigNumber(val), false);
                if (_n.isNaN()) {
                    /** @type {?} */
                    var minBigNum = new bignumber_js.BigNumber('' + this.min);
                    if (this.canNull || minBigNum.isNaN()) {
                        return null;
                    }
                    else {
                        _n = minBigNum;
                    }
                }
                return ( /** @type {?} */(this._getRealValue(_n)));
            };
        /**
         * @param {?} val
         * @param {?=} type
         * @return {?}
         */
        NumberRangeComponent.prototype.onModelChange = /**
         * @param {?} val
         * @param {?=} type
         * @return {?}
         */
            function (val, type) {
                if (type === 'begin') {
                    this._beginValue = val;
                }
                else if (type === 'end') {
                    this._endValue = val;
                }
                this.value = this.getRangeValue();
                this.onChangeCallback(this.value);
                this.onTouchedCallback();
                this.valueChange.emit(this.value);
            };
        /**
         * @param {?} $event
         * @return {?}
         */
        NumberRangeComponent.prototype.onClick = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                $event.stopPropagation();
            };
        /**
         * @param {?} e
         * @param {?} type
         * @return {?}
         */
        NumberRangeComponent.prototype.onBlur = /**
         * @param {?} e
         * @param {?} type
         * @return {?}
         */
            function (e, type) {
                if (!this.readonly && !this.disabled) {
                    /** @type {?} */
                    var tmpValue = e.value;
                    this.focusState = { isFocus: false, target: type };
                    if (type === 'begin') {
                        /** @type {?} */
                        var nextVal = this.getEndValue();
                        /** @type {?} */
                        var minValue = this.isEmpty(tmpValue)
                            ? ''
                            : this.isEmpty(nextVal)
                                ? tmpValue
                                : bignumber_js.BigNumber.min(tmpValue, nextVal).toFixed();
                        // this.prevValue = this.format(minValue);
                        this.beginValue = this.validInterval(new bignumber_js.BigNumber(minValue), false).toFixed();
                        // this.beginValue = this.getRealValue(this.prevValue);
                        this.inputPrev.nativeElement.value = this.prevValue;
                        this.onModelChange(this._beginValue, type);
                    }
                    else {
                        /** @type {?} */
                        var beginVal = this.getBeginValue();
                        /** @type {?} */
                        var maxValue = this.isEmpty(tmpValue)
                            ? ''
                            : this.isEmpty(beginVal)
                                ? tmpValue
                                : bignumber_js.BigNumber.max(tmpValue, beginVal).toFixed();
                        // this.nextValue = this.format(maxValue);
                        // this.endValue = this.getRealValue(this.nextValue);
                        this.endValue = this.validInterval(new bignumber_js.BigNumber(maxValue), false).toFixed();
                        this.inputNext.nativeElement.value = this.nextValue;
                        this.onModelChange(this._endValue, type);
                    }
                }
                this.valueChange.emit(this.value);
                if (type === 'begin') {
                    this.beginValueChange.emit(this.preNumValue);
                }
                else {
                    this.endValueChange.emit(this.nextNumValue);
                }
                this.blur.emit();
            };
        /**
         * @param {?=} type
         * @return {?}
         */
        NumberRangeComponent.prototype.onFocus = /**
         * @param {?=} type
         * @return {?}
         */
            function (type) {
                if (!this.readonly && !this.disabled) {
                    this.focusState = { isFocus: true, target: type };
                    if (type === 'begin') {
                        this.inputPrev.nativeElement.value = this._beginValue;
                        this.inputPrev.nativeElement.setSelectionRange(0, this._beginValue.length);
                        this.expendWidth('begin', 3);
                    }
                    else {
                        this.inputNext.nativeElement.value = this._endValue;
                        this.inputNext.nativeElement.setSelectionRange(0, this._endValue.length);
                        this.expendWidth('end', 3);
                    }
                }
                this.focus.emit(type);
            };
        /**
         * @param {?} which
         * @param {?} num
         * @return {?}
         */
        NumberRangeComponent.prototype.expendWidth = /**
         * @param {?} which
         * @param {?} num
         * @return {?}
         */
            function (which, num) {
                /** @type {?} */
                var target;
                /** @type {?} */
                var other;
                if (which === 'begin') {
                    target = this.prevWrapper.nativeElement;
                    other = this.nextWrapper.nativeElement;
                }
                else {
                    target = this.nextWrapper.nativeElement;
                    other = this.prevWrapper.nativeElement;
                }
                var width = target.getBoundingClientRect().width;
                if (width < 150) {
                    this.render.setStyle(target, 'flex', num);
                    this.render.setStyle(other, 'flex', 1);
                }
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NumberRangeComponent.prototype.onMouseEnter = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.isActiveTip = this.isShowPopover();
                if (this.nextValue && this.prevValue) {
                    this.popValue = this.prevValue + '~' + this.nextValue;
                }
                else {
                    this.popValue = '';
                }
            };
        /**
         * @return {?}
         */
        NumberRangeComponent.prototype.isShowPopover = /**
         * @return {?}
         */
            function () {
                var width = this.el.nativeElement.getBoundingClientRect().width;
                if (width && width < 200) {
                    return true;
                }
                return false;
            };
        /**
         * @param {?} $event
         * @param {?} typ
         * @return {?}
         */
        NumberRangeComponent.prototype.onKeyDownRange = /**
         * @param {?} $event
         * @param {?} typ
         * @return {?}
         */
            function ($event, typ) {
                if ($event.key === 'ArrowDown') {
                    $event.preventDefault();
                    this.down($event, typ);
                }
                if ($event.key === 'ArrowUp') {
                    $event.preventDefault();
                    this.up($event, typ);
                }
                $event.stopPropagation();
            };
        /**
         * @param {?} val
         * @return {?}
         */
        NumberRangeComponent.prototype.writeValue = /**
         * @param {?} val
         * @return {?}
         */
            function (val) {
                this.setValue(val);
            };
        /**
         * @param {?} val
         * @return {?}
         */
        NumberRangeComponent.prototype.setValue = /**
         * @param {?} val
         * @return {?}
         */
            function (val) {
                if (typeof val === 'string') {
                    /** @type {?} */
                    var valArr = val.split(this.separator);
                    this._beginValue = valArr[0] ? valArr[0] : '';
                    this._endValue = valArr[1] ? valArr[1] : '';
                }
                else {
                    this._beginValue = '';
                    this._beginValue = '';
                }
                this.prevValue = this.format(this._beginValue);
                this.nextValue = this.format(this._endValue);
            };
        //#region 上下操作按钮 或 上下方向键
        //#region 上下操作按钮 或 上下方向键
        /**
         * @param {?} e
         * @param {?} type
         * @return {?}
         */
        NumberRangeComponent.prototype.up =
            //#region 上下操作按钮 或 上下方向键
            /**
             * @param {?} e
             * @param {?} type
             * @return {?}
             */
            function (e, type) {
                e.stopPropagation();
                this._compute(type, 'up', (( /** @type {?} */(e.target))).value);
            };
        /**
         * @param {?} e
         * @param {?} type
         * @return {?}
         */
        NumberRangeComponent.prototype.down = /**
         * @param {?} e
         * @param {?} type
         * @return {?}
         */
            function (e, type) {
                e.stopPropagation();
                this._compute(type, 'down', (( /** @type {?} */(e.target))).value);
            };
        /**
         * @private
         * @param {?} type
         * @param {?} dir
         * @param {?} tagVal
         * @return {?}
         */
        NumberRangeComponent.prototype._compute = /**
         * @private
         * @param {?} type
         * @param {?} dir
         * @param {?} tagVal
         * @return {?}
         */
            function (type, dir, tagVal) {
                /** @type {?} */
                var m = dir === 'up' ? 'plus' : 'minus';
                // if (type === 'begin' && this.isDisableOfBtn(dir, this.prevValue)) {
                //     const bv = tagVal || 0;
                //     const ev = this.getRealValue(this.nextValue);
                //     const r = new BigNumber(bv)[m](Number(this.step));
                //     let v = r.toFixed();
                //     const minValue = BigNumber.min(v, ev);
                //     if (!minValue.eq(v) && !this.isEmpty(ev)) {
                //         v = ev;
                //     }
                //     this._beginValue = v;
                //     // this.prevValue = this.format(v);
                //     this.inputPrev.nativeElement.value = v;
                //     this.onModelChange(v, type);
                //     this.beginValueChange.emit(v);
                // } else if (type === 'end' && this.isDisableOfBtn(dir, this.nextValue)) {
                //     const bv = this.getRealValue(this.prevValue);
                //     const ev = tagVal || 0;
                //     const r = new BigNumber(ev)[m](Number(this.step));
                //     let v = r.toFixed();
                //     const maxValue = BigNumber.max(bv, v);
                //     if (!maxValue.eq(v) && !this.isEmpty(bv)) {
                //         v = bv;
                //     }
                //     this._endValue = v;
                //     // this.nextValue = this.format(v);
                //     this.inputNext.nativeElement.value = v;
                //     this.onModelChange(v, type);
                //     this.endValueChange.emit(v);
                // }
                if (type === 'begin' && this.isDisableOfBtn(dir, this.prevValue)) {
                    /** @type {?} */
                    var tmpValue = this.inputPrev.nativeElement.value;
                    /** @type {?} */
                    var nextVal = this.getEndValue();
                    /** @type {?} */
                    var bv = this.cleanNumString(tmpValue) || 0;
                    /** @type {?} */
                    var r = new bignumber_js.BigNumber(bv)[m](Number(this.step));
                    /** @type {?} */
                    var v = r.toFixed();
                    /** @type {?} */
                    var minValue = this.isEmpty(v)
                        ? ''
                        : this.isEmpty(nextVal)
                            ? v
                            : bignumber_js.BigNumber.min(v, nextVal).toFixed();
                    this.beginValue = this.validInterval(new bignumber_js.BigNumber(minValue), false).toFixed();
                    this.inputPrev.nativeElement.value = this.prevValue;
                    this.onModelChange(this._beginValue, type);
                    this.beginValueChange.emit(this._beginValue);
                }
                else if (type === 'end' && this.isDisableOfBtn(dir, this.nextValue)) {
                    /** @type {?} */
                    var tmpValue = this.inputNext.nativeElement.value;
                    /** @type {?} */
                    var beginVal = this.getBeginValue();
                    /** @type {?} */
                    var bv = this.cleanNumString(tmpValue) || 0;
                    /** @type {?} */
                    var r = new bignumber_js.BigNumber(bv)[m](Number(this.step));
                    /** @type {?} */
                    var v = r.toFixed();
                    /** @type {?} */
                    var maxValue = this.isEmpty(v)
                        ? ''
                        : this.isEmpty(beginVal)
                            ? v
                            : bignumber_js.BigNumber.max(v, beginVal).toFixed();
                    this.endValue = this.validInterval(new bignumber_js.BigNumber(maxValue), false).toFixed();
                    this.inputNext.nativeElement.value = this.nextValue;
                    this.onModelChange(this._endValue, type);
                    this.endValueChange.emit(this._endValue);
                }
                this.valueChange.emit(this.value);
                this.expendWidth(type, 3);
            };
        NumberRangeComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'farris-number-range',
                        template: "<div\r\n    class=\"input-group number-range f-cmp-number-spinner\"\r\n    [ngClass]=\"{\r\n        'f-state-disabled': disabled,\r\n        'f-state-readonly': readonly\r\n    }\"\r\n    (click)=\"$event.stopPropagation()\" (mouseenter)=\"onMouseEnter($event)\"\r\n>\r\n    <div class=\"form-control input-container\"\r\n        #popover=\"bs-popover\"\r\n        farrisPopover\r\n        [popover]=\"popValue\"\r\n        [triggers]=\"'hover'\"\r\n        [popActive]=\"isActiveTip\"\r\n        [showAction]=\"'show2'\"\r\n        [container]=\"'body'\"\r\n    >\r\n        <div class=\"sub-input-group\" #prevWrapper>\r\n            <input\r\n                type=\"text\"\r\n                number-only\r\n                [precision]=\"precision\"\r\n                autocomplete=\"off\"\r\n                class=\"sub-input form-control f-state-focus f-state-hover\"\r\n                [(ngModel)]=\"prevValue\"\r\n                #inputPrev\r\n                (blur)=\"onBlur(inputPrev, 'begin')\"\r\n                (focus)=\"onFocus('begin')\"\r\n                (ngModelChange)=\"onModelChange($event, 'begin')\"\r\n                [placeholder]=\"disabled || readonly ?'':beginPlaceHolder\"\r\n                [readOnly]=\"disabled || readonly ? 'readonly' : ''\"\r\n                [style.text-align]=\"textAlign\"\r\n                (keydown)=\"onKeyDownRange($event, 'begin')\"\r\n                (click)=\"onClick($event)\"\r\n            />\r\n            <div class=\"input-group-append btn-group btn-group-number sub-btn-group\" *ngIf=\"!readonly && !disabled && showButton\">\r\n                <span class=\"icon k-i-arrow-chevron-up number-arrow-chevron\" [style.cursor]=\"isDisableOfBtn('up', prevValue) ? 'pointer' : 'not-allowed'\"\r\n                    (click)=\"up($event, 'begin')\"></span>\r\n                <span class=\"icon k-i-arrow-chevron-down number-arrow-chevron\" [style.cursor]=\"isDisableOfBtn('down', prevValue) ? 'pointer' : 'not-allowed'\"\r\n                    (click)=\"down($event, 'begin')\"></span>\r\n            </div>\r\n        </div>\r\n        <span class=\"spliter\">~</span>\r\n        <div class=\"sub-input-group\" #nextWrapper>\r\n            <input\r\n                type=\"text\"\r\n                number-only\r\n                [precision]=\"precision\"\r\n                autocomplete=\"off\"\r\n                class=\"sub-input form-control f-state-focus f-state-hover\"\r\n                [(ngModel)]=\"nextValue\"\r\n                #inputNext\r\n                (blur)=\"onBlur(inputNext, 'end')\"\r\n                (focus)=\"onFocus('end')\"\r\n                (ngModelChange)=\"onModelChange($event, 'end')\"\r\n                [placeholder]=\"disabled || readonly ?'':endPlaceHolder\"\r\n                [readOnly]=\"disabled || readonly ? 'readonly' : ''\"\r\n                [style.text-align]=\"textAlign\"\r\n                (keydown)=\"onKeyDownRange($event, 'end')\"\r\n                (click)=\"onClick($event)\"\r\n            />\r\n            <div class=\"input-group-append btn-group btn-group-number sub-btn-group\" *ngIf=\"!readonly && !disabled  && showButton\"[style.text-align]=\"textAlign\" style=\"border-right:0\">\r\n                    <span class=\"icon k-i-arrow-chevron-up number-arrow-chevron\"  [style.cursor]=\"isDisableOfBtn('up', nextValue) ? 'pointer' : 'not-allowed'\"\r\n                        (click)=\"up($event, 'end')\"></span>\r\n                    <span class=\"icon k-i-arrow-chevron-down number-arrow-chevron\"\r\n                        [style.cursor]=\"isDisableOfBtn('down', nextValue) ? 'pointer' : 'not-allowed'\"\r\n                        (click)=\"down($event, 'end')\"></span>\r\n            </div>\r\n        </div>\r\n    </div>\r\n</div>\r\n",
                        providers: [
                            {
                                provide: forms.NG_VALUE_ACCESSOR,
                                useExisting: i0.forwardRef(( /**
                                 * @return {?}
                                 */function () { return NumberRangeComponent; })),
                                multi: true
                            }
                        ],
                        styles: [".input-group .btn-group-number{height:1.50003rem;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;background-color:#fff;border-left:1px solid #d9d9d9;width:1.50003rem}.input-group .btn-group-number .number-arrow-chevron{-webkit-box-flex:1;flex:1;line-height:1;height:50%;text-align:center}.input-group .btn-group-number .number-arrow-chevron:active{background-color:#e6e9f0}.input-group .btn-group-number .number-arrow-chevron:nth-child(2){border-top:1px solid #d9d9d9}input::-webkit-inner-spin-button,input::-webkit-outer-spin-button{-webkit-appearance:none!important;margin:0}.number-range{position:relative}.number-range .input-container{display:-webkit-box;display:flex;padding:0}.number-range .input-container .sub-input-group{-webkit-box-flex:1;flex:1;position:relative;display:-webkit-box;display:flex;-webkit-transition:.3s ease-out;transition:.3s ease-out}.number-range .input-container .sub-input-group .sub-input{width:100%;border:none;outline:0;background-color:rgba(0,0,0,0);min-width:2px;padding:.125rem 4px .125rem .5rem}.number-range .input-container .spliter{width:15px;text-align:center}"]
                    }] }
        ];
        NumberRangeComponent.propDecorators = {
            beginPlaceHolder: [{ type: i0.Input }],
            endPlaceHolder: [{ type: i0.Input }],
            inputPrev: [{ type: i0.ViewChild, args: ['inputPrev',] }],
            inputNext: [{ type: i0.ViewChild, args: ['inputNext',] }],
            prevWrapper: [{ type: i0.ViewChild, args: ['prevWrapper',] }],
            nextWrapper: [{ type: i0.ViewChild, args: ['nextWrapper',] }],
            popover: [{ type: i0.ViewChild, args: ['popover',] }],
            separator: [{ type: i0.Input }],
            beginValue: [{ type: i0.Input }],
            endValue: [{ type: i0.Input }],
            value: [{ type: i0.Input }],
            beginValueChange: [{ type: i0.Output }],
            endValueChange: [{ type: i0.Output }]
        };
        return NumberRangeComponent;
    }(NumberSpinnerComponent));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NumberSpinnerModule = /** @class */ (function () {
        function NumberSpinnerModule() {
        }
        NumberSpinnerModule.decorators = [
            { type: i0.NgModule, args: [{
                        declarations: [
                            NumberSpinnerComponent, NumberOnlyDirective,
                            NumberRangeComponent
                        ],
                        imports: [
                            common.CommonModule, forms.FormsModule,
                            uiPopover.PopoverModule.forRoot(),
                            uiLocale.LocaleModule.forRoot()
                        ],
                        exports: [NumberSpinnerComponent, NumberOnlyDirective,
                            NumberRangeComponent]
                    },] }
        ];
        return NumberSpinnerModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.NumberSpinnerService = NumberSpinnerService;
    exports.NumberSpinnerComponent = NumberSpinnerComponent;
    exports.NumberSpinnerModule = NumberSpinnerModule;
    exports.NumberOnlyDirective = NumberOnlyDirective;
    exports.NumberRangeComponent = NumberRangeComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
