// @gsp-lcm/dbo-selector/esm5/lib/dbo-select.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DboSelectComponent } from './dbo-select/dbo-select.component';
import { GridModule } from '@progress/kendo-angular-grid';
import { FormsModule } from '@angular/forms';
import { DboHelpDevService } from './dbo-help-dev.service';
import { HttpClientModule } from '@angular/common/http';
var DboSelectModule = /** @class */ (function () {
    function DboSelectModule() {
    }
    DboSelectModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        GridModule,
                        FormsModule,
                        HttpClientModule
                    ],
                    providers: [DboHelpDevService],
                    declarations: [DboSelectComponent],
                    entryComponents: [DboSelectComponent],
                    exports: [DboSelectComponent]
                },] }
    ];
    return DboSelectModule;
}());
export { DboSelectModule };
