// @gsp-wf/wf-process-model/esm2015/lib/model/extension/tool/Component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { Tool } from "./Tool";
export class Component extends Tool {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = Component.Clr_Type_ID;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new Component(this.Model);
    }
    //Json 序列化
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "MetadataId", this.MetadataId);
        BpmnModelHelper.SetAttributeValue(obj, "ComponentType", this.ComponentType);
        BpmnModelHelper.SetAttributeValue(obj, "MethodCode", this.MethodCode);
        BpmnModelHelper.SetAttributeValue(obj, "ScopeType", this.ScopeType);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.MetadataId = obj.MetadataId;
        this.ComponentType = obj.ComponentType;
        this.MethodCode = obj.MethodCode;
        this.ScopeType = obj.ScopeType;
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const component = new Component(this.Model);
        component.Id = BpmnModelHelper.GenerateElementId();
        component.Code = this.Code;
        component.Name = this.Name;
        component.Description = this.Description;
        component.Execution = this.Execution;
        component.AsyncCallBackType = this.AsyncCallBackType;
        component.Time = this.Time;
        component.isForCompensation = this.isForCompensation;
        if (this.ActualParameters && this.ActualParameters.length > 0) {
            for (const p of this.ActualParameters) {
                /** @type {?} */
                const parameter = p.clone();
                component.ActualParameters.push(parameter);
            }
        }
        component.MetadataId = this.MetadataId;
        component.ComponentType = this.ComponentType;
        component.MethodCode = this.MethodCode;
        component.ScopeType = this.ScopeType;
        return component;
    }
}
Component.Clr_Type_ID = "wf.Component";
if (false) {
    /** @type {?} */
    Component.Clr_Type_ID;
    /** @type {?} */
    Component.prototype.ClrTypeID;
    /** @type {?} */
    Component.prototype.MetadataId;
    /** @type {?} */
    Component.prototype.ComponentType;
    /** @type {?} */
    Component.prototype.MethodCode;
    /** @type {?} */
    Component.prototype.ScopeType;
}
BpmnModelHelper.RegisterElementType(new Component(new BpmnModel()));
