// @gsp-wf/wf-process-model/esm2015/lib/model/extension/tool/ToolTime.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const ToolTime = {
    /// <summary> 过程实例已经创建事件 </summary>
    PROCESSINSTANCE_CREATED: "PROCESSINSTANCE_CREATED",
    /// <summary> 过程实例已经启动事件 </summary>
    PROCESSINSTANCE_STARTED: "PROCESSINSTANCE_STARTED",
    /// <summary> 过程实例已经挂起事件 </summary>
    PROCESSINSTANCE_SUSPENDED: "PROCESSINSTANCE_SUSPENDED",
    /// <summary> 过程实例已经暂停事件 </summary>
    PROCESSINSTANCE_RESUMED: "PROCESSINSTANCE_RESUMED",
    /// <summary> 过程实例已经完成事件 </summary>  
    PROCESSINSTANCE_COMPLETED: "PROCESSINSTANCE_COMPLETED",
    /// <summary> 过程实例已经中止事件 </summary>
    PROCESSINSTANCE_TERMINATED: "PROCESSINSTANCE_TERMINATED",
    ///<summary>流程实例预警启动事件</summary>
    PROCESSINSTANCE_ALARM: "PROCESSINSTANCE_ALARM",
    ///<summary>流程实例超时事件</summary> 
    PROCESSINSTANCE_TIMEOUT: "PROCESSINSTANCE_TIMEOUT",
    /// <summary> 过程实例已经删除事件 </summary>
    PROCESSINSTANCE_DELETED: "PROCESSINSTANCE_DELETED",
    PROCESSINSTANCE_REACTIVATED: 'PROCESSINSTANCE_REACTIVATED',
    PROCESSINSTANCE_CANCELED: 'PROCESSINSTANCE_CANCELED',
    PROCESSINSTANCE_ROLLBACK: 'PROCESSINSTANCE_ROLLBACK',
    /// <summary> 驳回后重新提交 </summary>
    PROCESSINSTANCE_RESUBMIT: 'PROCESSINSTANCE_RESUBMIT',
    /// <summary> 环节办理人变更事件 </summary>
    ACTIVITY_AND_PARTICIPANT_UPDATED: "ACTIVITY_AND_PARTICIPANT_UPDATED",
    /// <summary> 活动实例已经创建事件 </summary>
    ACTIVITY_CREATED: "ACTIVITY_CREATED",
    /// <summary> 活动实例已经启动事件 </summary>
    ACTIVITY_STARTED: "ACTIVITY_STARTED",
    /// <summary> 活动实例已经挂起事件 </summary>
    ACTIVITY_SUSPENDED: "ACTIVITY_SUSPENDED",
    /// <summary> 活动实例已经恢复事件 </summary>
    ACTIVITY_RESUMED: "ACTIVITY_RESUMED",
    /// <summary> 活动实例已经完成事件 </summary>
    ACTIVITY_COMPLETED: "ACTIVITY_COMPLETED",
    /// <summary> 活动实例已经终止事件 </summary>
    ACTIVITY_TERMINATED: "ACTIVITY_TERMINATED",
    /// <summary> 工作项已经创建事件 </summary>
    WORKITEM_CREATED: "WORKITEM_CREATED",
    /// <summary> 工作项已经分配事件 </summary>
    WORKITEM_ASSIGNED: "WORKITEM_ASSIGNED",
    /// <summary> 工作项已经启动事件 </summary>
    WORKITEM_STARTED: "WORKITEM_STARTED",
    /// <summary> 工作项已经挂起事件 </summary>
    WORKITEM_SUSPENDED: "WORKITEM_SUSPENDED",
    /// <summary> 工作项已经暂停事件 </summary>
    WORKITEM_RESUMED: "WORKITEM_RESUMED",
    /// <summary> 工作项已经完成事件 </summary>
    WORKITEM_COMPLETED: "WORKITEM_COMPLETED",
    /// <summary> 工作项已经终止事件 </summary>
    WORKITEM_TERMINATED: "WORKITEM_TERMINATED",
    /// <summary> 工作项预警启动事件 </summary>
    WORKITEM_ALARM: "WORKITEM_ALARM",
    /// <summary> 工作项超时事件</summary>
    WORKITEM_TIMEOUT: "WORKITEM_TIMEOUT",
    /// <summary> 未定义</summary>
    UNDEFINITION: "UNDEFINITION",
    LOADDATA: "LOADDATA",
};
export { ToolTime };
