// @gsp-wf/wf-process-model/esm2015/lib/model/extension/tool/Tool.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { AsyncCallBackType } from "../../misc/AsyncCallBackType";
import { ExecutionType } from "../../misc/ExecutionType";
import { ActualParameter } from "../../parameter/WFActualParameter";
import { WFExtensionElement } from "../WFExtensionElement";
import { ToolTime } from "./ToolTime";
/**
 * @abstract
 */
export class Tool extends WFExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = Tool.Clr_Type_ID;
        this.ActualParameters = new Array();
        this.Execution = ExecutionType.SYNCHRONOUS;
        this.AsyncCallBackType = AsyncCallBackType.NO;
        this.Time = ToolTime.UNDEFINITION;
        this.isForCompensation = false;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        /** @type {?} */
        let actualParametersElements = [];
        if (this.ActualParameters !== null && this.ActualParameters.length > 0) {
            this.ActualParameters.forEach((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                actualParametersElements.push(item.ConvertToJsonObject());
            }));
        }
        BpmnModelHelper.SetAttributeValue(obj, "ActualParameters", actualParametersElements);
        BpmnModelHelper.SetAttributeValue(obj, 'Execution', this.Execution);
        BpmnModelHelper.SetAttributeValue(obj, 'Time', this.Time);
        BpmnModelHelper.SetAttributeValue(obj, 'IsForCompensation', this.isForCompensation);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        /** @type {?} */
        let actualParameters = (/** @type {?} */ (obj.ActualParameters));
        if (actualParameters !== null && actualParameters.length > 0) {
            actualParameters.forEach((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                /** @type {?} */
                let actualParameter = new ActualParameter(this.Model);
                actualParameter.LoadFromJsonObject(item);
                this.ActualParameters.push(actualParameter);
            }));
        }
        this.Execution = (/** @type {?} */ (obj.Execution));
        this.Time = (/** @type {?} */ (obj.Time));
        this.isForCompensation = obj.IsForCompensation;
    }
}
Tool.Clr_Type_ID = "wf.Tool";
if (false) {
    /** @type {?} */
    Tool.Clr_Type_ID;
    /** @type {?} */
    Tool.prototype.ClrTypeID;
    /** @type {?} */
    Tool.prototype.ActualParameters;
    /** @type {?} */
    Tool.prototype.Execution;
    /** @type {?} */
    Tool.prototype.AsyncCallBackType;
    /** @type {?} */
    Tool.prototype.Time;
    /** @type {?} */
    Tool.prototype.isForCompensation;
}
