// @qdp/query-framework/esm5/lib/component/condition-dialog-tab/condition-dialog-tab.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, HostBinding, ViewChild } from '@angular/core';
import { FilterExpressionComponent } from '@gsp-svc/filtercondition';
import { QDPCacheUtil, RtfServices } from '@qdp/common';
var ConditionDialogTabComponent = /** @class */ (function () {
    function ConditionDialogTabComponent() {
        this.cls = '';
        this.tabIndex = 0;
        this.controlIndexChanged = 0;
        // injector: Injector;
        // @ViewChild('adHost', { read: ViewContainerRef }) adHost: ViewContainerRef;
        // componentFactoryResolver: ComponentFactoryResolver;
        // filterInstance: any;
        this.propertyList = [];
        this.filterExpressionString = '';
        this.showFilterExpression = true;
        this.showForm = false;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ConditionDialogTabComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        for (var changeName in changes) {
            /** @type {?} */
            var newValue = changes[changeName].currentValue;
            if (newValue !== null && newValue !== void 0) {
                switch (changeName) {
                    case 'propertyList':
                        if (newValue) {
                            this.propertyList = tslib_1.__spread(newValue);
                        }
                        break;
                    case 'filterExpressionString':
                        if (newValue) {
                            this.filterExpressionString = newValue;
                        }
                        break;
                }
            }
        }
    };
    /**
     * @return {?}
     */
    ConditionDialogTabComponent.prototype.changeCss = /**
     * @return {?}
     */
    function () {
        this.cls = this.tabIndex === this.controlIndexChanged ? 'f-utils-fill' : '';
    };
    /**
     * @return {?}
     */
    ConditionDialogTabComponent.prototype.createFilterComponent = /**
     * @return {?}
     */
    function () {
        // if (this.injector) {
        //   this.componentFactoryResolver = this.injector.get(ComponentFactoryResolver);
        //   const componentFactory = this.componentFactoryResolver.resolveComponentFactory(FilterExpressionComponent);
        //   this.adHost.clear();
        //   const componentRef = this.adHost.createComponent(componentFactory);
        //   this.filterInstance = componentRef.instance;
        //   this.filterInstance.propertyList = this.propertyList;
        //   this.filterInstance.filterExpressionString = this.filterExpressionString;
        // }
    };
    /**
     * @return {?}
     */
    ConditionDialogTabComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        try {
            if (QDPCacheUtil.getCache(RtfServices.getTabId('condition-dialog-tabs') + 'condition-dialog-tabs')) {
                QDPCacheUtil.getCache(RtfServices.getTabId('condition-dialog-tabs') + 'condition-dialog-tabs').push(this);
            }
            else {
                QDPCacheUtil.setCache(RtfServices.getTabId('condition-dialog-tabs') + 'condition-dialog-tabs', [this]);
            }
        }
        catch (e) {
        }
        this.changeCss();
    };
    ConditionDialogTabComponent.decorators = [
        { type: Component, args: [{
                    selector: 'condition-dialog-tab',
                    template: "<div class=\"h-100 f-utils-fill\" [style.display]=\"tabIndex===controlIndexChanged?'block':'none'\">\r\n  <div class=\"f-utils-fill-flex-row h-100\" *ngIf=\"tabIndex==1 && showFilterExpression\">\r\n    <div class=\"f-utils-fill-auto h-100\">\r\n      <app-filter-expression #funcFilter class=\"h-100\" [propertyList]=\"propertyList\" [filterExpressionString]=\"filterExpressionString\"></app-filter-expression>\r\n    </div>\r\n  </div>\r\n  <div class=\"f-utils-fill-flex-row h-100\" *ngIf=\"tabIndex!=1 || showForm\">\r\n    <div class=\"f-utils-fill-auto h-100\">\r\n      <ng-content></ng-content>\r\n    </div>\r\n  </div>\r\n</div>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ConditionDialogTabComponent.ctorParameters = function () { return []; };
    ConditionDialogTabComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        tabIndex: [{ type: Input }],
        filterInstance: [{ type: ViewChild, args: [FilterExpressionComponent,] }],
        propertyList: [{ type: Input }],
        filterExpressionString: [{ type: Input }],
        showFilterExpression: [{ type: Input }],
        showForm: [{ type: Input }]
    };
    return ConditionDialogTabComponent;
}());
export { ConditionDialogTabComponent };
if (false) {
    /** @type {?} */
    ConditionDialogTabComponent.prototype.cls;
    /** @type {?} */
    ConditionDialogTabComponent.prototype.tabIndex;
    /** @type {?} */
    ConditionDialogTabComponent.prototype.controlIndexChanged;
    /** @type {?} */
    ConditionDialogTabComponent.prototype.filterInstance;
    /** @type {?} */
    ConditionDialogTabComponent.prototype.propertyList;
    /** @type {?} */
    ConditionDialogTabComponent.prototype.filterExpressionString;
    /** @type {?} */
    ConditionDialogTabComponent.prototype.showFilterExpression;
    /** @type {?} */
    ConditionDialogTabComponent.prototype.showForm;
}
