// @gsp-cmp/ccmp-ui/esm2015/lib/ccmp-ui.rt-service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, EventEmitter, ComponentFactoryResolver } from '@angular/core';
import { CommonComponentSelectorComponent } from './ccmp-selector/common-component-selector.component';
import { ViewUtils } from './view-utils/view-utils';
import { BsModalService } from '@farris/ui-modal';
import { HttpHeaders, HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { ComponentInfo } from './params/component-info';
import * as i0 from "@angular/core";
import * as i1 from "./view-utils/view-utils";
import * as i2 from "@farris/ui-modal";
import * as i3 from "@angular/common/http";
export class CommonComponentUiRTService {
    /**
     * @param {?} resolver
     * @param {?} injector
     * @param {?} util
     * @param {?} modalService
     * @param {?} http
     */
    constructor(resolver, injector, util, modalService, http) {
        this.resolver = resolver;
        this.injector = injector;
        this.util = util;
        this.modalService = modalService;
        this.http = http;
    }
    /**
     * 通用构件操作选择器
     * @param {?} metadataId 通用构件元数据
     * @param {?} sessionId 功能sessionId
     * @param {?=} options UI可选配置参数
     * @return {?}
     */
    openSelector(metadataId, sessionId, options) {
        return this.getComponentInfo(metadataId, sessionId).pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            if (data.obselete) {
                this.util.showObseleteDialog(data);
                return { confirmEvent: new EventEmitter(), closeEvent: new EventEmitter() };
            }
            else {
                return this.create(data, options);
            }
        })));
    }
    /**
     * 获取构件信息
     * @param {?} metadataId 构件元数据ID
     * @param {?} sessionId 功能sessionid
     * @return {?}
     */
    getComponentInfo(metadataId, sessionId) {
        /** @type {?} */
        let headers = new HttpHeaders();
        headers.set("sessionId", sessionId);
        /** @type {?} */
        let url = `/api/runtime/sys/v1.0/common-component/component-info/${metadataId}`;
        return this.http.get(url, { headers: headers }).pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            if (data) {
                /** @type {?} */
                let cmpInfo = new ComponentInfo();
                cmpInfo.init(data);
                return cmpInfo;
            }
        })));
    }
    /**
     * @private
     * @param {?} componentInfo
     * @param {?} options
     * @return {?}
     */
    create(componentInfo, options) {
        /** @type {?} */
        let fac = this.resolver.resolveComponentFactory(CommonComponentSelectorComponent);
        /** @type {?} */
        let cmpRef = fac.create(this.injector);
        cmpRef.instance.componentInfo = componentInfo;
        /** @type {?} */
        let modalConfig = this.util.getModalConfig(cmpRef.instance.btnRef, options);
        /** @type {?} */
        let dialog = this.modalService.show(cmpRef, modalConfig);
        cmpRef.instance.dialog = dialog;
        return { confirmEvent: cmpRef.instance.confirmEmitter, closeEvent: cmpRef.instance.closeEmitter };
    }
}
CommonComponentUiRTService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
CommonComponentUiRTService.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: Injector },
    { type: ViewUtils },
    { type: BsModalService },
    { type: HttpClient }
];
/** @nocollapse */ CommonComponentUiRTService.ngInjectableDef = i0.defineInjectable({ factory: function CommonComponentUiRTService_Factory() { return new CommonComponentUiRTService(i0.inject(i0.ComponentFactoryResolver), i0.inject(i0.INJECTOR), i0.inject(i1.ViewUtils), i0.inject(i2.BsModalService), i0.inject(i3.HttpClient)); }, token: CommonComponentUiRTService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    CommonComponentUiRTService.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    CommonComponentUiRTService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    CommonComponentUiRTService.prototype.util;
    /**
     * @type {?}
     * @private
     */
    CommonComponentUiRTService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    CommonComponentUiRTService.prototype.http;
}
