// @gsp-wf/wf-process-editor/esm5/lib/views/vo-config/vo-config.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ComponentFactoryResolver, EventEmitter, Inject, Injector, Input, LOCALE_ID, Optional, Output, TemplateRef, ViewChild } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { LoadingService } from '@farris/ui-loading';
import { MessagerService } from '@farris/ui-messager';
import { BsModalService } from '@farris/ui-modal';
import { NotifyService } from '@farris/ui-notify';
import { SelectMetadataRTComponent } from '@gsp-lcm/metadatart-selector';
import { FrameworkService } from '@gsp-sys/rtf-common';
import { BillTypeHelpComponent } from '../bill-type-help/bill-type-help/bill-type-help.component';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../i18n/pipe/wf-process-editor-locale.pipe';
var VoConfigComponent = /** @class */ (function () {
    function VoConfigComponent(messager, notify, modal, framework, router, activatedRoute, loading, cfr, inj, cdr, localeId, resources, localePipe) {
        this.messager = messager;
        this.notify = notify;
        this.modal = modal;
        this.framework = framework;
        this.router = router;
        this.activatedRoute = activatedRoute;
        this.loading = loading;
        this.cfr = cfr;
        this.inj = inj;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.groupIcon = '<i class="f-icon f-icon-lookup"></i>';
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.editorParams = {};
        this.modalConfig = {
            title: this.localePipe.transform("component.voConfig.documentInfo"),
            width: 500,
            height: 240,
            showButtons: true
        };
        this.entity = {};
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
    }
    Object.defineProperty(VoConfigComponent.prototype, "value", {
        set: /**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                this.entity = data || {};
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    VoConfigComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    VoConfigComponent.prototype.clickOk = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = { value: this.entity, parameters: null };
        this.submitModal.emit(obj);
    };
    /**
     * @return {?}
     */
    VoConfigComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeModal.emit();
    };
    /**
     * @return {?}
     */
    VoConfigComponent.prototype.selectVo = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var compFactory = this.cfr.resolveComponentFactory(SelectMetadataRTComponent);
        /** @type {?} */
        var compRef = compFactory.create(this.inj);
        compRef.instance.typeList = 'GSPViewModel';
        // 元数据选择器bizTypeId表示业务种类Id
        compRef.instance.bizTypeId = this.editorParams && this.editorParams.bizTypeId;
        compRef.instance.selectedMetadata.subscribe((/**
         * @param {?} metadata4RefDto
         * @return {?}
         */
        function (metadata4RefDto) {
            _this.clearVo();
            /** @type {?} */
            var voDto = metadata4RefDto.metadata;
            /** @type {?} */
            var vo = JSON.parse(voDto.content);
            _this.entity.currentVoId = vo.ID;
            _this.entity.currentVoName = vo.Name;
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            console.error(err);
            // todo 确定异常结构, 目前采用兼容取值方式
            _this.messager.error(err.error == null ? err.message : (err.error.Message || err.error.message || err.message));
            _this.clearVo();
        }));
        /** @type {?} */
        var modalConfig = {
            title: this.localePipe.transform("component.voConfig.viewModelSelection"),
            width: 930,
            height: 573,
            showButtons: false
        };
        /** @type {?} */
        var dialog = this.modal.show(compRef, modalConfig);
        compRef.instance.closeModalEmitter.subscribe((/**
         * @param {?} $event
         * @return {?}
         */
        function ($event) { return dialog.close(); }));
    };
    /**
     * @return {?}
     */
    VoConfigComponent.prototype.clearVo = /**
     * @return {?}
     */
    function () {
        this.entity.currentVoId = undefined;
        this.entity.currentVoName = undefined;
    };
    /**
     * @return {?}
     */
    VoConfigComponent.prototype.selectDocuType = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var cmpF = this.cfr.resolveComponentFactory(BillTypeHelpComponent);
        /** @type {?} */
        var cmpR = cmpF.create(this.inj, []);
        cmpR.instance.docuTypeId = this.entity.currentDocumentTypeId;
        /** @type {?} */
        var buttons = [];
        buttons.push({
            text: this.localePipe.transform("component.voConfig.cancel"),
            cls: "btn btn-secondary",
            handle: (/**
             * @return {?}
             */
            function () {
                bsModal && bsModal.close();
            })
        });
        buttons.push({
            text: this.localePipe.transform("component.voConfig.clickOk"),
            cls: "btn btn-primary",
            handle: (/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var docuType = cmpR.instance.getSelectedDocuType();
                if (docuType == null) {
                    _this.notify.warning(_this.localePipe.transform("component.voConfig.selectOne"));
                    return;
                }
                _this.entity.currentDocumentTypeId = docuType.id;
                _this.entity.currentDocumentTypeName = docuType.name;
                bsModal && bsModal.close();
            })
        });
        /** @type {?} */
        var bsModal = this.modal.show(cmpR, {
            title: this.localePipe.transform("component.voConfig.documentTypeSelect"),
            height: 573,
            width: 930,
            buttons: buttons,
            showMaxButton: false,
        });
    };
    /**
     * @return {?}
     */
    VoConfigComponent.prototype.clearDocuType = /**
     * @return {?}
     */
    function () {
        this.entity.currentDocumentTypeId = undefined;
        this.entity.currentDocumentTypeName = undefined;
    };
    VoConfigComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-vo-config',
                    template: "<div class=\"f-page f-page-card f-page-is-mainsubcard\" id=\"bif-manager-list\">\r\n  <div class=\"f-page-main\" id=\"page-main\" style=\"overflow-y: auto;\">\r\n    <div class=\"f-form-layout farris-form farris-form-controls-inline\">\r\n      <div class=\"col-12\">\r\n        <div class=\"farris-group-wrap\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\" title=\"{{'component.voConfig.viewModel' | wfLocale}}\" for=\"vo\">\r\n              <span class=\"farris-label-text\">{{'component.voConfig.viewModel' | wfLocale}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input-group id=\"vo\" [ngModel]=\"entity&&entity.currentVoName\" [editable]=\"false\" [enableClear]=\"false\"\r\n                (clickHandle)=\"selectVo()\" (clear)=\"clearVo()\" [groupText]=\"groupIcon\" placeholder=\"{{'component.voConfig.placeholder' | wfLocale}}\">\r\n              </input-group>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n\r\n      <div class=\"col-12\">\r\n        <div class=\"farris-group-wrap\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\" title=\"{{'component.voConfig.documenttype' | wfLocale}}\" for=\"vo\">\r\n              <span class=\"farris-label-text\">{{'component.voConfig.documenttype' | wfLocale}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input-group id=\"vo\" [ngModel]=\"entity&&entity.currentDocumentTypeName\" [editable]=\"false\"\r\n                [enableClear]=\"false\" (clickHandle)=\"selectDocuType()\" (clear)=\"clearDocuType()\" [groupText]=\"groupIcon\"\r\n                placeholder=\"{{'component.voConfig.placeholder' | wfLocale}}\">\r\n              </input-group>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm px-3 mx-2\" (click)=\"clickCancel()\">{{'component.voConfig.cancel' | wfLocale}}</button>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm px-3 mx-2\" (click)=\"clickOk()\">{{'component.voConfig.clickOk' | wfLocale}}</button>\r\n</ng-template>\r\n",
                    providers: [
                        WfLocalePipe,
                    ]
                }] }
    ];
    /** @nocollapse */
    VoConfigComponent.ctorParameters = function () { return [
        { type: MessagerService },
        { type: NotifyService },
        { type: BsModalService },
        { type: FrameworkService },
        { type: Router },
        { type: ActivatedRoute },
        { type: LoadingService },
        { type: ComponentFactoryResolver },
        { type: Injector },
        { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
        { type: WfLocalePipe, decorators: [{ type: Optional }] }
    ]; };
    VoConfigComponent.propDecorators = {
        closeModal: [{ type: Output }],
        submitModal: [{ type: Output }],
        value: [{ type: Input }],
        editorParams: [{ type: Input }],
        modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }]
    };
    return VoConfigComponent;
}());
export { VoConfigComponent };
if (false) {
    /** @type {?} */
    VoConfigComponent.prototype.groupIcon;
    /** @type {?} */
    VoConfigComponent.prototype.closeModal;
    /** @type {?} */
    VoConfigComponent.prototype.submitModal;
    /** @type {?} */
    VoConfigComponent.prototype.editorParams;
    /** @type {?} */
    VoConfigComponent.prototype.modalFooter;
    /** @type {?} */
    VoConfigComponent.prototype.modalConfig;
    /** @type {?} */
    VoConfigComponent.prototype.entity;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.messager;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.modal;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.framework;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.loading;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.inj;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.localePipe;
}
var VoConfigConverter = /** @class */ (function () {
    function VoConfigConverter(localePipe) {
        this.localePipe = localePipe;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    VoConfigConverter.prototype.convertTo = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data && data.currentVoId) {
            return data.currentVoName || data.currentVoId || this.localePipe.transform("component.voConfig.set");
        }
        return this.localePipe.transform("component.voConfig.notSet");
    };
    return VoConfigConverter;
}());
export { VoConfigConverter };
if (false) {
    /**
     * @type {?}
     * @private
     */
    VoConfigConverter.prototype.localePipe;
}
