// @qdp/formular/esm2015/lib/directive/help-before-event.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directive/help-before-event.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input } from '@angular/core';
import { LookupGridComponent } from '@farris/ui-lookup';
import { RestfulService } from '../service/restful.service';
import { of } from 'rxjs';
import { tap, switchMap } from 'rxjs/operators';
export class HelpBeforeEventDirective {
    /**
     * @param {?} lookup
     * @param {?} restfulService
     */
    constructor(lookup, restfulService) {
        this.lookup = lookup;
        this.restfulService = restfulService;
        this.helpData = {};
        this.localHost = {};
        this.dynamicInput = [];
        this.dictPicking = (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const options = {
                headers: {
                    'Content-Type': 'application/json'
                },
                responseType: 'text'
            };
            /** @type {?} */
            const paramvalues = {};
            this.dynamicInput.forEach((/**
             * @param {?} currentInput
             * @return {?}
             */
            currentInput => {
                paramvalues[currentInput.paramcode] = currentInput.paramvalue;
            }));
            /** @type {?} */
            const params = {
                helpId: this.helpData.fieldConfig.helpId,
                filterInfo: this.helpData.fieldConfig.filterInfo,
                parameters: JSON.stringify(paramvalues)
            };
            /** @type {?} */
            const getHelpDataUrl = this.localHost + '/api/runtime/bcc/v1.0/qdpsearch/getHelpData';
            /** @type {?} */
            const r$ = of({
                showDialog: true
            });
            return this.restfulService.post(getHelpDataUrl, params, {}, options).pipe(tap((/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                this.lookup.items = JSON.parse(data);
            })), switchMap((/**
             * @return {?}
             */
            () => r$)));
        });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        console.log(this.helpData, this.lookup);
        this.lookup.dictPicking = this.dictPicking;
        this.lookup.dictPicked = (/**
         * @param {?} rowData
         * @return {?}
         */
        (rowData) => {
            this.helpData.paramvalue = rowData[this.helpData.fieldConfig.valueField] + '|' + rowData[this.helpData.fieldConfig.textField];
            /** @type {?} */
            const r$ = of({
                closeDialog: true
            });
            return r$;
        });
    }
}
HelpBeforeEventDirective.decorators = [
    { type: Directive, args: [{
                selector: '[libHelpBeforeEvent]'
            },] }
];
/** @nocollapse */
HelpBeforeEventDirective.ctorParameters = () => [
    { type: LookupGridComponent },
    { type: RestfulService }
];
HelpBeforeEventDirective.propDecorators = {
    helpData: [{ type: Input, args: ["libHelpBeforeEvent",] }],
    localHost: [{ type: Input }],
    dynamicInput: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    HelpBeforeEventDirective.prototype.helpData;
    /** @type {?} */
    HelpBeforeEventDirective.prototype.localHost;
    /** @type {?} */
    HelpBeforeEventDirective.prototype.dynamicInput;
    /** @type {?} */
    HelpBeforeEventDirective.prototype.dictPicking;
    /**
     * @type {?}
     * @private
     */
    HelpBeforeEventDirective.prototype.lookup;
    /**
     * @type {?}
     * @private
     */
    HelpBeforeEventDirective.prototype.restfulService;
}
