// @gsp-wf/wf-process-editor/esm2015/lib/views/performer/views/matrix-performer/service/matrix-performer.component.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { map } from 'rxjs/operators';
export class MatrixDefHelpService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @param {?} url
     * @param {?=} params
     * @return {?}
     */
    getData(url, params) {
        if (url.indexOf('getMatrixDef') > -1) {
            /** @type {?} */
            const index = params.pageIndex || 1;
            /** @type {?} */
            const size = params.pageSize || 20;
            /** @type {?} */
            const searchValue = JSON.parse(params.searchValue);
            if (searchValue.searchValue) {
                /** @type {?} */
                let query = '{';
                if (searchValue.searchValue) {
                    if (searchValue.searchField === '*') {
                        query += `"codeOrNameLike":"${searchValue.searchValue}",`;
                    }
                    else {
                        query += `"${searchValue.searchField}":"${searchValue.searchValue}",`;
                    }
                }
                if (query.substr(query.length - 1, 1) === ',') {
                    query = query.substring(0, query.length - 1);
                }
                query += '}';
                query = encodeURIComponent(query);
                query += `&pageIndex=${index}&pageSize=${size}`;
                url += '?searchValue=' + query;
            }
            else {
                url += `?pageIndex=${index}&pageSize=${size}`;
            }
            return this.http.get(url).pipe(map((/**
             * @param {?} re
             * @return {?}
             */
            (re) => ({
                items: re.content,
                total: re.totalElements,
                pageInfo: {
                    pageIndex: re.pageable.pageNumber + 1,
                    pageSize: re.pageable.pageSize,
                    pageList: [20, 40, 60, 80, 100],
                    enablePager: true
                }
            }))));
        }
    }
}
MatrixDefHelpService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
MatrixDefHelpService.ctorParameters = () => [
    { type: HttpService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    MatrixDefHelpService.prototype.http;
}
