// @qdp/formular/esm5/lib/component/step-component/step-component.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component/step-component/step-component.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewContainerRef, ComponentFactoryResolver, Output, Input, ViewChild, EventEmitter } from '@angular/core';
import { find, findIndex, merge } from 'lodash-es';
import { LocalizeService } from '@qdp/localize';
var StepComponentComponent = /** @class */ (function () {
    function StepComponentComponent(viewContainerRef, componentFactoryResolver, localizeService) {
        this.viewContainerRef = viewContainerRef;
        this.componentFactoryResolver = componentFactoryResolver;
        this.localizeService = localizeService;
        this.clickCancelBtn = new EventEmitter();
        this.clickFinishBtn = new EventEmitter();
        this.title = '';
        this.steps = [];
        this._copySteps = [];
        this.curStepIndex = 0;
        this.labelPre = '上一步';
        this.labelNext = '下一步';
        this.labelFinish = '完成';
        this.labelDelete = '删除';
        this.labelPre = this.localizeService.getValue('formular.step.labelPre');
        this.labelNext = this.localizeService.getValue('formular.step.labelNext');
        this.labelFinish = this.localizeService.getValue('formular.step.labelFinish');
        this.labelDelete = this.localizeService.getValue('formular.step.labelDelete');
    }
    /**
     * @return {?}
     */
    StepComponentComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.steps && this.steps.length > 0) {
            this.curStep = this.steps[this.curStepIndex];
            this.createStepComponent(this.steps[this.curStepIndex]);
            // this.createStepComponent(this.steps[4]);
        }
        this._copySteps = this.steps;
    };
    /**
     * @param {?} step
     * @return {?}
     */
    StepComponentComponent.prototype.createStepComponent = /**
     * @param {?} step
     * @return {?}
     */
    function (step) {
        var _this = this;
        this.stepContainer.clear();
        /** @type {?} */
        var componentFactory = this.componentFactoryResolver.resolveComponentFactory(step.component);
        this.curComponentRef = this.stepContainer.createComponent(componentFactory);
        this.curComponentRef.instance.data = step.data;
        // 新增步骤
        if (this.curComponentRef.instance.addStep) {
            this.curComponentRef.instance.addStep.subscribe((/**
             * @param {?} stepList
             * @return {?}
             */
            function (stepList) {
                var _a;
                /** @type {?} */
                var curIndex = findIndex(_this.steps, _this.curStep);
                (_a = _this.steps).splice.apply(_a, tslib_1.__spread([curIndex + 1, 0], stepList));
                _this.curStep.nextStep = stepList[0].stepId;
                _this.curStep.showFinishBtn = false;
                _this.curStep.showNextStepBtn = true;
            }));
        }
        // 删除步骤
        if (this.curComponentRef.instance.removeStep) {
            this.curComponentRef.instance.removeStep.subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                merge(_this.curStep, data.curStep);
                data.removeSteps.forEach((/**
                 * @param {?} removeStep
                 * @return {?}
                 */
                function (removeStep) {
                    /** @type {?} */
                    var index = findIndex(_this.steps, removeStep);
                    if (index >= 0) {
                        _this.steps.splice(index, 1);
                    }
                }));
            }));
        }
        // 变更步骤
        if (this.curComponentRef.instance.changeStep) {
            this.curComponentRef.instance.changeStep.subscribe((/**
             * @param {?} stepList
             * @return {?}
             */
            function (stepList) {
                stepList.forEach((/**
                 * @param {?} changeStep
                 * @return {?}
                 */
                function (changeStep) {
                    /** @type {?} */
                    var oldStep = find(_this._copySteps, (/**
                     * @param {?} s
                     * @return {?}
                     */
                    function (s) { return s.stepId === changeStep.stepId; }));
                    merge(oldStep, changeStep);
                }));
                _this.steps = _this._copySteps.filter((/**
                 * @param {?} step
                 * @return {?}
                 */
                function (step) { return step.show !== false; }));
            }));
        }
    };
    /**
     * @param {?} nextStep
     * @return {?}
     */
    StepComponentComponent.prototype.jumpStep = /**
     * @param {?} nextStep
     * @return {?}
     */
    function (nextStep) {
        // 判断跳转方向
        if (nextStep === this.curStep) {
            return;
        }
        /** @type {?} */
        var nextIndex = findIndex(this.steps, (/**
         * @param {?} s
         * @return {?}
         */
        function (s) { return s.stepId === nextStep.stepId; }));
        if (nextIndex === this.curStepIndex) {
            return;
        }
        if (nextIndex < this.curStepIndex) { // 向前跳
            this.curStep = nextStep;
            this.curStepIndex = nextIndex;
            this.createStepComponent(nextStep);
            return;
        }
        if (nextStep.stepId === this.curStep.nextStep) { // 下一步
            this.nextStep();
            return;
        }
        //  else if (this.finishedSteps.indexOf(nextStep.stepId) > -1) { // 已完成的后续步骤
        //   this.curStep = nextStep;
        //   this.curStepIndex = nextIndex;
        //   this.createStepComponent(nextStep);
        // }
    };
    /**
      * 点击上一步
      */
    /**
     * 点击上一步
     * @return {?}
     */
    StepComponentComponent.prototype.lastStep = /**
     * 点击上一步
     * @return {?}
     */
    function () {
        /** @type {?} */
        var self = this;
        this.curStepIndex = findIndex(this.steps, (/**
         * @param {?} s
         * @return {?}
         */
        function (s) { return s.nextStep === self.curStep.stepId; }));
        this.curStep = this.steps[this.curStepIndex];
        // 执行各Component定义的nextStepClick事件
        /** @type {?} */
        var clickLastStep = this.curComponentRef.instance.clickLastStep;
        if (clickLastStep && typeof (clickLastStep) === 'function') {
            /** @type {?} */
            var result = clickLastStep();
            if (typeof (result) === 'boolean' && result) {
                this.createStepComponent(this.curStep);
            }
        }
    };
    /**
     * 点击下一步
     */
    /**
     * 点击下一步
     * @return {?}
     */
    StepComponentComponent.prototype.nextStep = /**
     * 点击下一步
     * @return {?}
     */
    function () {
        var _this = this;
        // 校验
        /** @type {?} */
        var validationFunc = this.curComponentRef.instance.validation;
        if (validationFunc && typeof (validationFunc) === 'function') {
            /** @type {?} */
            var result = validationFunc();
            if (!result) {
                return;
            }
        }
        // 执行各Component定义的nextStepClick事件
        /** @type {?} */
        var clickNextStep = this.curComponentRef.instance.clickNextStep;
        if (clickNextStep && typeof (clickNextStep) === 'function') {
            /** @type {?} */
            var result = clickNextStep();
            if (typeof (result) === 'boolean' && result) {
                this.moveToNextStep(result);
            }
            else {
                if (result) {
                    result.subscribe((/**
                     * @param {?} data
                     * @return {?}
                     */
                    function (data) {
                        _this.moveToNextStep(data);
                    }));
                }
            }
        }
        else {
            this.moveToNextStep(true);
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    StepComponentComponent.prototype.moveToNextStep = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (!data) {
            return;
        }
        /** @type {?} */
        var curStep = this.curStep;
        if (curStep.nextStep === '') {
            return;
        }
        // this.finishedSteps.push(curStep.stepId);
        this.curStepIndex = findIndex(this.steps, (/**
         * @param {?} s
         * @return {?}
         */
        function (s) { return s.stepId === curStep.nextStep; }));
        if (this.curStepIndex > -1) {
            this.curStep = this.steps[this.curStepIndex];
            this.createStepComponent(this.curStep);
        }
    };
    /**
     * 取消
     */
    /**
     * 取消
     * @return {?}
     */
    StepComponentComponent.prototype.clickCancel = /**
     * 取消
     * @return {?}
     */
    function () {
        this.clickCancelBtn.emit();
    };
    /**
     * 完成
     */
    /**
     * 完成
     * @return {?}
     */
    StepComponentComponent.prototype.finishWizard = /**
     * 完成
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var validationFunc = this.curComponentRef.instance.validation;
        if (validationFunc && typeof (validationFunc) === 'function') {
            /** @type {?} */
            var result = validationFunc();
            if (!result) {
                return;
            }
        }
        /** @type {?} */
        var beforeFinish = this.curComponentRef.instance.beforeFinish;
        if (this.curComponentRef.instance.qdpResult && !beforeFinish) {
            beforeFinish = this.curComponentRef.instance.qdpResult.beforeFinish;
        }
        if (beforeFinish && typeof (beforeFinish) === 'function') {
            /** @type {?} */
            var result = beforeFinish();
            if (typeof (result) === 'boolean' && result) {
                this.clickFinishBtn.emit();
            }
            else if (typeof (result) !== 'boolean') {
                if (result) {
                    result.subscribe((/**
                     * @param {?} data
                     * @return {?}
                     */
                    function (data) {
                        if (data) {
                            _this.clickFinishBtn.emit(data);
                        }
                    }));
                }
            }
        }
    };
    StepComponentComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-step-component',
                    template: "<div class=\"stepWizard\">\r\n  <div class=\"createNav d-flex flex-wrap\" *ngIf=\"steps.length>1\">\r\n    <div *ngFor=\"let step of steps;let i = index\" class=\"d-flex step\" [class.active]=\"curStepIndex>=i\">\r\n      <div class=\"d-flex\" *ngIf=\"step.show\" (click)=\"jumpStep(step)\">\r\n        <div class=\"pointer d-flex\">\r\n          <div class=\"stepIndex\"><span>{{i+1}}</span></div>\r\n          <div class=\"stepTitle\">{{step.title}}</div>\r\n        </div>\r\n        <div class=\"triangle_border_right\" *ngIf=\"step.nextStep\">\r\n          <span>\u00B7\u00B7\u00B7\u00B7\u00B7></span>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n\r\n  <div class=\"content\">\r\n    <perfect-scrollbar [config]=\"{suppressScrollX: true}\">\r\n      <ng-container #stepContainer></ng-container>\r\n    </perfect-scrollbar>\r\n  </div>\r\n\r\n\r\n\r\n  <div class=\"footBtns\">\r\n    <button type=\"button\" class=\"btn btn-secondary btn-sm px-3 mx-2\" *ngIf=\"curStep.showLastStepBtn\" (click)=\"lastStep()\"\r\n      type=\"submit\">{{labelPre}}</button>\r\n    <button type=\"button\" class=\"btn btn-primary btn-sm px-3 mx-2\" *ngIf=\"curStep.showNextStepBtn\" (click)=\"nextStep()\"\r\n      type=\"submit\">{{labelNext}}</button>\r\n    <button type=\"button\" class=\"btn btn-primary btn-sm px-3 mx-2\" *ngIf=\"curStep.showFinishBtn\" (click)=\"finishWizard()\">{{labelFinish}}</button>\r\n  </div>\r\n</div>\r\n",
                    styles: [":host{position:absolute;top:0;bottom:0;left:0;right:0}.stepWizard{height:100%;display:flex;flex-direction:column;background:#f8f8fa}.stepWizard .createNav{background:#fff;box-shadow:0 2px 2px 0 rgba(0,0,0,.04);height:50px;overflow:hidden;padding:0 27px}.stepWizard .createNav div{margin:auto 0}.stepWizard .createNav .pointer{cursor:pointer}.stepWizard .createNav .stepIndex{height:24px;width:24px;background-color:#d9d9d9;color:#fff;border-radius:100%;text-align:center}.stepWizard .createNav .step.active .stepIndex{background-color:#59a1ff}.stepWizard .createNav .stepIndex span{line-height:24px}.stepWizard .createNav .step.active .stepIndex span{background-color:#59a1ff}.stepWizard .createNav .stepTitle{color:#8c8c8c;padding-left:11px;font-size:14px}.stepWizard .createNav .step.active .stepTitle{color:rgba(0,0,0,.85)}.stepWizard .createNav .stepTitle.disable{color:#cdcdcf}.stepWizard .createNav .step .triangle_border_right{position:relative}.stepWizard .createNav .step .triangle_border_right span{padding:0 9px;color:#d9d9d9}.stepWizard .createNav .step.active .triangle_border_right span{color:#59a1ff}.stepWizard .content{flex:1;overflow:auto;font-size:.875rem}.stepWizard .content .ps-content{width:inherit;height:inherit}.stepWizard .footBtns{width:100%;display:flex;align-items:center;justify-content:flex-end;padding:10px 1rem;background:#f4f4f4;border-top:1px solid #ddd}"]
                }] }
    ];
    /** @nocollapse */
    StepComponentComponent.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: ComponentFactoryResolver },
        { type: LocalizeService }
    ]; };
    StepComponentComponent.propDecorators = {
        clickCancelBtn: [{ type: Output }],
        clickFinishBtn: [{ type: Output }],
        title: [{ type: Input }],
        steps: [{ type: Input }],
        stepContainer: [{ type: ViewChild, args: ['stepContainer', { read: ViewContainerRef },] }]
    };
    return StepComponentComponent;
}());
export { StepComponentComponent };
if (false) {
    /** @type {?} */
    StepComponentComponent.prototype.clickCancelBtn;
    /** @type {?} */
    StepComponentComponent.prototype.clickFinishBtn;
    /** @type {?} */
    StepComponentComponent.prototype.title;
    /** @type {?} */
    StepComponentComponent.prototype.steps;
    /** @type {?} */
    StepComponentComponent.prototype._copySteps;
    /** @type {?} */
    StepComponentComponent.prototype.curStep;
    /** @type {?} */
    StepComponentComponent.prototype.curStepIndex;
    /** @type {?} */
    StepComponentComponent.prototype.curComponentRef;
    /** @type {?} */
    StepComponentComponent.prototype.stepContainer;
    /** @type {?} */
    StepComponentComponent.prototype.labelPre;
    /** @type {?} */
    StepComponentComponent.prototype.labelNext;
    /** @type {?} */
    StepComponentComponent.prototype.labelFinish;
    /** @type {?} */
    StepComponentComponent.prototype.labelDelete;
    /**
     * @type {?}
     * @private
     */
    StepComponentComponent.prototype.viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    StepComponentComponent.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    StepComponentComponent.prototype.localizeService;
}
