// @gsp-aif/runtime-api/esm2015/lib/aif-pull-select-creation-rule/service/aif-pull-select-creation-rule.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, ComponentFactoryResolver } from "@angular/core";
import { AifPullSelectCreationRuleComponent } from "../aif-pull-select-creation-rule/aif-pull-select-creation-rule.component";
import { BsModalService } from '@farris/ui-modal';
/**
 * 供外部调用的拉式选择生单规则服务类接口
 */
export class AifPullSelectCreationRuleService {
    /**
     * @param {?} injector
     * @param {?} cfr
     * @param {?} modalService
     */
    constructor(injector, cfr, modalService) {
        this.injector = injector;
        this.cfr = cfr;
        this.modalService = modalService;
        this.innerModalOptions = {
            title: '生单规则选择',
            width: 650,
            height: 415,
            buttons: null,
            showButtons: false,
        };
    }
    /**
     * 显示选择生单规则模态框，点击确定按钮时返回选中的生单规则
     * @param {?} args 查询生单规则所需参数
     * @param {?=} modalOptions 模态框配置,可为空
     * @return {?}
     */
    showCrRuleSelectModal(args, modalOptions) {
        /** @type {?} */
        const cmpF = this.cfr.resolveComponentFactory(AifPullSelectCreationRuleComponent);
        /** @type {?} */
        const cmpR = cmpF.create(this.injector);
        cmpR.instance.getCreationRule(args);
        cmpR.instance.bsModalRef = this.modalService.show(cmpR, this.isNullOrUndefined(modalOptions) ? this.innerModalOptions : modalOptions);
        return cmpR.instance.subject.asObservable();
    }
    /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    isNullOrUndefined(arg) {
        return arg == null;
    }
}
AifPullSelectCreationRuleService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AifPullSelectCreationRuleService.ctorParameters = () => [
    { type: Injector },
    { type: ComponentFactoryResolver },
    { type: BsModalService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AifPullSelectCreationRuleService.prototype.innerModalOptions;
    /**
     * @type {?}
     * @private
     */
    AifPullSelectCreationRuleService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    AifPullSelectCreationRuleService.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    AifPullSelectCreationRuleService.prototype.modalService;
}
