// @farris/ide-devkit/esm2015/lib/components/modules.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { BsDropdownModule } from './dropdown';
import { NgbTabsetModule } from './tabset/tabset.module';
import { NgbNavTabsetModule } from './navtab/navtab.module';
import { NgbAccordionModule } from './accordion/accordion.module';
import { PanelModule } from './panel/panel.module';
/** @type {?} */
const MODULES = [
    BsDropdownModule
];
/** @type {?} */
const NGB_MODULES = [
    NgbTabsetModule,
    NgbNavTabsetModule,
    NgbAccordionModule
];
/** @type {?} */
const IDE_MODULES = [
    PanelModule
];
// @NgModule({
//   imports: [
//     BsDropdownModule.forRoot()
//   ],
//   exports: [BsDropdownModule],
//   schemas: [NO_ERRORS_SCHEMA]
// })
// export class MDBRootModule {
// }
// @NgModule({ exports: [BsDropdownModule] })
// export class MDBBootstrapModule {
//   public static forRoot(): ModuleWithProviders {
//     return { ngModule: MDBRootModule };
//   }
// }
export class NgbRootModule {
}
NgbRootModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    NgbTabsetModule.forRoot(),
                    NgbNavTabsetModule.forRoot(),
                    NgbAccordionModule.forRoot()
                ],
                exports: [
                    NGB_MODULES
                ]
            },] }
];
export class NgbModule {
    /**
     * @return {?}
     */
    static forRoot() { return { ngModule: NgbRootModule }; }
}
NgbModule.decorators = [
    { type: NgModule, args: [{},] }
];
export class IDERootModule {
}
IDERootModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    PanelModule.forRoot()
                ]
            },] }
];
