// @gsp-wf/wf-sign/fesm5/gsp-wf-wf-sign.js
import sp from 'signature_pad/dist/signature_pad.umd';
import { Injectable, Component, ViewChild, ElementRef, Input, Output, EventEmitter, NgModule, defineInjectable } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/wf-sign.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var WfSignService = /** @class */ (function () {
    function WfSignService() {
    }
    WfSignService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    WfSignService.ctorParameters = function () { return []; };
    /** @nocollapse */ WfSignService.ngInjectableDef = defineInjectable({ factory: function WfSignService_Factory() { return new WfSignService(); }, token: WfSignService, providedIn: "root" });
    return WfSignService;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/wf-sign.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var WfSignComponent = /** @class */ (function () {
    function WfSignComponent() {
    }
    /**
     * @return {?}
     */
    WfSignComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    WfSignComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-sign-wf-sign',
                    template: "\n    <p>\n      wf-sign works!\n    </p>\n  "
                }] }
    ];
    /** @nocollapse */
    WfSignComponent.ctorParameters = function () { return []; };
    return WfSignComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/signature/signature.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SignatureComponent = /** @class */ (function () {
    function SignatureComponent(elementRef) {
        // no op
        this.elementRef = elementRef;
        this.options = this.options || {};
        this.onBeginEvent = new EventEmitter();
        this.onEndEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    SignatureComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var canvas = this.elementRef.nativeElement.querySelector('canvas');
        if (((/** @type {?} */ (this.options))).canvasHeight) {
            canvas.height = ((/** @type {?} */ (this.options))).canvasHeight;
        }
        if (((/** @type {?} */ (this.options))).canvasWidth) {
            canvas.width = ((/** @type {?} */ (this.options))).canvasWidth;
        }
        this.signaturePad = new sp(canvas, this.options);
        this.signaturePad.onBegin = this.onBegin.bind(this);
        this.signaturePad.onEnd = this.onEnd.bind(this);
    };
    /**
     * @return {?}
     */
    SignatureComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var canvas = this.elementRef.nativeElement.querySelector('canvas');
        canvas.width = 0;
        canvas.height = 0;
    };
    /**
     * @return {?}
     */
    SignatureComponent.prototype.resizeCanvas = /**
     * @return {?}
     */
    function () {
        // When zoomed out to less than 100%, for some very strange reason,
        // some browsers report devicePixelRatio as less than 1
        // and only part of the canvas is cleared then.
        /** @type {?} */
        var ratio = Math.max(window.devicePixelRatio || 1, 1);
        //const canvas: any = this.signaturePad._canvas;
        /** @type {?} */
        var canvas = this.signaturePad.canvas;
        canvas.width = canvas.offsetWidth * ratio;
        canvas.height = canvas.offsetHeight * ratio;
        canvas.getContext('2d').scale(ratio, ratio);
        this.signaturePad.clear(); // otherwise isEmpty() might return incorrect value
    };
    /**
     * @param {?} text
     * @return {?}
     */
    SignatureComponent.prototype.setWaterMark = /**
     * @param {?} text
     * @return {?}
     */
    function (text) {
        /** @type {?} */
        var canvas = this.elementRef.nativeElement.querySelector('canvas');
        /** @type {?} */
        var ctx = canvas.getContext("2d");
        ctx.font = "20px 微软雅黑";
        ctx.fillStyle = '#888';
        ctx.fillText(text, 10, 50);
    };
    //
    //
    /**
     * @return {?}
     */
    SignatureComponent.prototype.setImageBackground = 
    //
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var canvas = this.elementRef.nativeElement.querySelector('canvas');
        /** @type {?} */
        var ctx = canvas.getContext("2d");
        /** @type {?} */
        var imageData = ctx.getImageData(0, 0, ((/** @type {?} */ (this.options))).canvasWidth, ((/** @type {?} */ (this.options))).canvasHeight);
        for (var i = 0; i < imageData.data.length; i += 4) {
            // 当该像素是透明的，则设置成白色
            if (imageData.data[i + 3] == 0) {
                imageData.data[i] = 255;
                imageData.data[i + 1] = 255;
                imageData.data[i + 2] = 255;
                imageData.data[i + 3] = 255;
            }
        }
        ctx.putImageData(imageData, 0, 0);
    };
    /**
     * @param {?} image
     * @return {?}
     */
    SignatureComponent.prototype.setBackgroundImage = /**
     * @param {?} image
     * @return {?}
     */
    function (image) {
        /** @type {?} */
        var canvas = this.elementRef.nativeElement.querySelector('canvas');
        /** @type {?} */
        var ctx = canvas.getContext("2d");
        ctx.drawImage(image, 0, 0);
    };
    // Returns signature image as an array of point groups
    // Returns signature image as an array of point groups
    /**
     * @return {?}
     */
    SignatureComponent.prototype.toData = 
    // Returns signature image as an array of point groups
    /**
     * @return {?}
     */
    function () {
        if (this.signaturePad) {
            return this.signaturePad.toData();
        }
        else {
            return [];
        }
    };
    // Draws signature image from an array of point groups
    // Draws signature image from an array of point groups
    /**
     * @param {?} points
     * @return {?}
     */
    SignatureComponent.prototype.fromData = 
    // Draws signature image from an array of point groups
    /**
     * @param {?} points
     * @return {?}
     */
    function (points) {
        this.signaturePad.fromData(points);
    };
    // Returns signature image as data URL (see https://mdn.io/todataurl for the list of possible paramters)
    // Returns signature image as data URL (see https://mdn.io/todataurl for the list of possible paramters)
    /**
     * @param {?=} imageType
     * @param {?=} quality
     * @return {?}
     */
    SignatureComponent.prototype.toDataURL = 
    // Returns signature image as data URL (see https://mdn.io/todataurl for the list of possible paramters)
    /**
     * @param {?=} imageType
     * @param {?=} quality
     * @return {?}
     */
    function (imageType, quality) {
        return this.signaturePad.toDataURL(imageType, quality); // save image as data URL
    };
    // Draws signature image from data URL
    // Draws signature image from data URL
    /**
     * @param {?} dataURL
     * @param {?=} options
     * @return {?}
     */
    SignatureComponent.prototype.fromDataURL = 
    // Draws signature image from data URL
    /**
     * @param {?} dataURL
     * @param {?=} options
     * @return {?}
     */
    function (dataURL, options) {
        if (options === void 0) { options = {}; }
        // set default height and width on read data from URL
        if (!options.hasOwnProperty('height') && ((/** @type {?} */ (this.options))).canvasHeight) {
            options.height = ((/** @type {?} */ (this.options))).canvasHeight;
        }
        if (!options.hasOwnProperty('width') && ((/** @type {?} */ (this.options))).canvasWidth) {
            options.width = ((/** @type {?} */ (this.options))).canvasWidth;
        }
        this.signaturePad.fromDataURL(dataURL, options);
    };
    // Clears the canvas
    // Clears the canvas
    /**
     * @return {?}
     */
    SignatureComponent.prototype.clear = 
    // Clears the canvas
    /**
     * @return {?}
     */
    function () {
        this.signaturePad.clear();
    };
    // Returns true if canvas is empty, otherwise returns false
    // Returns true if canvas is empty, otherwise returns false
    /**
     * @return {?}
     */
    SignatureComponent.prototype.isEmpty = 
    // Returns true if canvas is empty, otherwise returns false
    /**
     * @return {?}
     */
    function () {
        return this.signaturePad.isEmpty();
    };
    // Unbinds all event handlers
    // Unbinds all event handlers
    /**
     * @return {?}
     */
    SignatureComponent.prototype.off = 
    // Unbinds all event handlers
    /**
     * @return {?}
     */
    function () {
        this.signaturePad.off();
    };
    // Rebinds all event handlers
    // Rebinds all event handlers
    /**
     * @return {?}
     */
    SignatureComponent.prototype.on = 
    // Rebinds all event handlers
    /**
     * @return {?}
     */
    function () {
        this.signaturePad.on();
    };
    // set an option on the signaturePad - e.g. set('minWidth', 50);
    // set an option on the signaturePad - e.g. set('minWidth', 50);
    /**
     * @param {?} option
     * @param {?} value
     * @return {?}
     */
    SignatureComponent.prototype.set = 
    // set an option on the signaturePad - e.g. set('minWidth', 50);
    /**
     * @param {?} option
     * @param {?} value
     * @return {?}
     */
    function (option, value) {
        switch (option) {
            case 'canvasHeight':
                this.signaturePad.canvas.height = value;
                break;
            case 'canvasWidth':
                this.signaturePad.canvas.width = value;
                break;
            default:
                this.signaturePad[option] = value;
        }
    };
    // notify subscribers on signature begin
    // notify subscribers on signature begin
    /**
     * @return {?}
     */
    SignatureComponent.prototype.onBegin = 
    // notify subscribers on signature begin
    /**
     * @return {?}
     */
    function () {
        this.onBeginEvent.emit(true);
    };
    // notify subscribers on signature end
    // notify subscribers on signature end
    /**
     * @return {?}
     */
    SignatureComponent.prototype.onEnd = 
    // notify subscribers on signature end
    /**
     * @return {?}
     */
    function () {
        this.onEndEvent.emit(true);
    };
    /**
     * @return {?}
     */
    SignatureComponent.prototype.queryPad = /**
     * @return {?}
     */
    function () {
        return this.signaturePad;
    };
    SignatureComponent.decorators = [
        { type: Component, args: [{
                    template: '<canvas></canvas>',
                    selector: 'wf-sign-signature'
                }] }
    ];
    /** @nocollapse */
    SignatureComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    SignatureComponent.propDecorators = {
        options: [{ type: Input }],
        onBeginEvent: [{ type: Output }],
        onEndEvent: [{ type: Output }]
    };
    return SignatureComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/sign/sign.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SignComponent = /** @class */ (function () {
    function SignComponent() {
        // 抛出选中的事件
        this.toDataUrlEvent = new EventEmitter();
        this.isDrawEvent = new EventEmitter();
        this.imageType = "image/png";
        this.imageQuality = 1;
        this.isDraw = false;
        this.signaturePadOptions = {
            // passed through to szimek/signature_pad constructor
            'minWidth': 1,
            'canvasWidth': 500,
            'canvasHeight': 240
        };
    }
    /**
     * @return {?}
     */
    SignComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    SignComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        // this.signaturePad is now available
        this.signaturePad.set('minWidth', 1); // set szimek/signature_pad options at runtime
        this.signaturePad.set('canvasWidth', this.canvasDiv.nativeElement.clientWidth);
        this.signaturePad.clear(); // invoke functions from szimek/signature_pad API
        this.signaturePad.setImageBackground();
    };
    /**
     * @return {?}
     */
    SignComponent.prototype.confirm = /**
     * @return {?}
     */
    function () {
        console.log(this.signaturePad.toDataURL(this.imageType, this.imageQuality));
        this.signaturePad.setWaterMark(new Date());
        this.toDataUrlEvent.emit(this.signaturePad.toDataURL(this.imageType, this.imageQuality));
    };
    /**
     * @return {?}
     */
    SignComponent.prototype.reset = /**
     * @return {?}
     */
    function () {
        this.signaturePad.clear();
        this.isDraw = false;
        this.isDrawEvent.emit(this.isDraw);
    };
    /**
     * @return {?}
     */
    SignComponent.prototype.drawComplete = /**
     * @return {?}
     */
    function () {
        this.isDraw = true;
        this.isDrawEvent.emit(this.isDraw);
        // will be notified of szimek/signature_pad's onEnd event
        //console.log(this.signaturePad.toDataURL());
    };
    /**
     * @return {?}
     */
    SignComponent.prototype.drawStart = /**
     * @return {?}
     */
    function () {
        // will be notified of szimek/signature_pad's onBegin event
        console.log('begin drawing');
    };
    SignComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-sign-sign',
                    template: "<div class=\"fm-feature-suggestion fm-feature-section--gutter\">\r\n    <div #canvastitle style=\"padding:2px 2px 2px 0;text-align:left;width:100%;height: 21px;\">\r\n        <span style=\"font-size: 15px;color: #999999;\">\u7B7E\u540D\u533A\u57DF</span>\r\n      </div>\r\n      <wf-sign-signature  [options]=\"signaturePadOptions\" (onBeginEvent)=\"drawStart()\" (onEndEvent)=\"drawComplete()\"></wf-sign-signature>\r\n  <div class=\"fm-feature-suggestion--bar\">\r\n    <ul class=\"suggestion--bar-common\">\r\n      <li class=\"suggestion--bar-item\" (click)=\"confirm()\">\r\n        <span>\u786E\u8BA4</span>\r\n      </li>\r\n      <li class=\"suggestion--bar-item\" (click)=\"reset()\">\r\n        <span>\u91CD\u7F6E</span>\r\n      </li>\r\n    </ul>\r\n  </div>\r\n</div>\r\n",
                    styles: [".fm-page-main1{overflow:hidden}.fm-utils-flex-column{display:flex;flex-direction:column}.fm-page{position:absolute;top:0;left:0;bottom:0;right:0;display:flex;flex-direction:column;background:#fff;font-size:13px;line-height:18px}.fm-page-main{flex-basis:0;flex-shrink:1;flex-grow:1;overflow-x:hidden;overflow-y:auto;position:relative}.fm-feature-section--gutter{margin:0 0 12px}.fm-feature-suggestion{background:#fff;padding:12px 16px}.fm-feature-suggestion .fm-feature-suggestion--bar{display:flex;flex-direction:row;align-items:center;color:#888}"]
                }] }
    ];
    /** @nocollapse */
    SignComponent.ctorParameters = function () { return []; };
    SignComponent.propDecorators = {
        signaturePad: [{ type: ViewChild, args: [SignatureComponent,] }],
        canvasDiv: [{ type: ViewChild, args: ['canvastitle',] }],
        toDataUrlEvent: [{ type: Output }],
        isDrawEvent: [{ type: Output }],
        imageType: [{ type: Input }],
        imageQuality: [{ type: Input }]
    };
    return SignComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/sign-here/sign-here.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SignHereComponent = /** @class */ (function () {
    function SignHereComponent() {
        // 抛出选中的事件
        this.imageType = "image/png";
        this.imageQuality = 1;
        this.title = '请对审批意见进行签字确认';
        this.isDraw = false;
        this.signaturePadOptions = {
            // passed through to szimek/signature_pad constructor
            'minWidth': 1,
            'canvasWidth': 500,
            'canvasHeight': 240
        };
    }
    /**
     * @return {?}
     */
    SignHereComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    SignHereComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        // this.signaturePad is now available
        this.signaturePad.set('minWidth', 1); // set szimek/signature_pad options at runtime
        this.signaturePad.set('canvasWidth', this.canvasDiv.nativeElement.clientWidth);
        this.signaturePad.clear(); // invoke functions from szimek/signature_pad API
        this.signaturePad.setImageBackground();
        if (this.image) {
            /** @type {?} */
            var img = new Image();
            img.src = this.image;
            this.signaturePad.setBackgroundImage(img);
        }
    };
    /**
     * @return {?}
     */
    SignHereComponent.prototype.confirm = /**
     * @return {?}
     */
    function () {
        console.log(this.signaturePad.toDataURL(this.imageType, this.imageQuality));
        //this.signaturePad.setWaterMark(new Date());
        return this.signaturePad.toDataURL(this.imageType, this.imageQuality);
    };
    /**
     * @return {?}
     */
    SignHereComponent.prototype.reset = /**
     * @return {?}
     */
    function () {
        this.signaturePad.clear();
        this.signaturePad.setImageBackground();
        this.isDraw = false;
    };
    /**
     * @return {?}
     */
    SignHereComponent.prototype.drawComplete = /**
     * @return {?}
     */
    function () {
        this.isDraw = true;
        // will be notified of szimek/signature_pad's onEnd event
        //console.log(this.signaturePad.toDataURL());
    };
    /**
     * @return {?}
     */
    SignHereComponent.prototype.drawStart = /**
     * @return {?}
     */
    function () {
        // will be notified of szimek/signature_pad's onBegin event
        console.log('begin drawing');
    };
    //设置背景图片
    //设置背景图片
    /**
     * @param {?} image
     * @return {?}
     */
    SignHereComponent.prototype.initCanvasByImage = 
    //设置背景图片
    /**
     * @param {?} image
     * @return {?}
     */
    function (image) {
        this.signaturePad.setBackgroundImage(image);
    };
    SignHereComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-sign-here',
                    template: "<div class=\"fm-feature-suggestion fm-feature-section--gutter\">\r\n    <div style=\"padding:2px 2px 2px 0;text-align:left;width:100%;height: 21px;\">\r\n        <span style=\"font-size: 15px;color: #999999;\">{{title}}</span>\r\n      </div>\r\n      <div class=\"wf-sign-signature-docker\">\r\n      <div #canvastitle class=\"wf-sign-signature\">\r\n      <wf-sign-signature  [options]=\"signaturePadOptions\" (onBeginEvent)=\"drawStart()\" (onEndEvent)=\"drawComplete()\"></wf-sign-signature>\r\n    </div>\r\n</div>\r\n</div>\r\n",
                    styles: [".fm-page-main1{overflow:hidden}.fm-utils-flex-column{display:flex;flex-direction:column}.fm-page{position:absolute;top:0;left:0;bottom:0;right:0;display:flex;flex-direction:column;background:#fff;font-size:13px;line-height:18px}.fm-page-main{flex-basis:0;flex-shrink:1;flex-grow:1;overflow-x:hidden;overflow-y:auto;position:relative}.fm-feature-section--gutter{margin:0 0 12px}.fm-feature-suggestion{background:#fff;padding:12px 16px}.fm-feature-suggestion .fm-feature-suggestion--bar{display:flex;flex-direction:row;align-items:center;color:#888}.wf-sign-signature{margin:1px}.wf-sign-signature-docker{border:1px dashed #ddd;margin-left:4px;margin-right:4px}"]
                }] }
    ];
    /** @nocollapse */
    SignHereComponent.ctorParameters = function () { return []; };
    SignHereComponent.propDecorators = {
        signaturePad: [{ type: ViewChild, args: [SignatureComponent,] }],
        canvasDiv: [{ type: ViewChild, args: ['canvastitle',] }],
        imageType: [{ type: Input }],
        imageQuality: [{ type: Input }],
        title: [{ type: Input }],
        image: [{ type: Input }]
    };
    return SignHereComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/wf-sign.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var WfSignModule = /** @class */ (function () {
    function WfSignModule() {
    }
    WfSignModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [WfSignComponent, SignatureComponent, SignComponent, SignHereComponent],
                    imports: [],
                    exports: [WfSignComponent, SignatureComponent, SignComponent, SignHereComponent]
                },] }
    ];
    return WfSignModule;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: gsp-wf-wf-sign.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { WfSignService, WfSignComponent, WfSignModule, SignHereComponent, SignComponent as ɵb, SignatureComponent as ɵa };
