// @gsp-wf/wf-process-model/esm5/lib/model/extension/completion-strategy/BackWardRule.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ExtensionElement, BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { TargetRefConstants } from "../../misc/TargetRefConstants";
var BackWardRule = /** @class */ (function (_super) {
    tslib_1.__extends(BackWardRule, _super);
    function BackWardRule(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = BackWardRule.Clr_Type_ID;
        _this.TargetRef = TargetRefConstants.INITIAL_ACTIVITY_ID;
        _this.SelectTargetAble = false;
        _this.WaitReturn = false;
        _this.ResubmitEffect = null;
        _this.DynamicBackWardRuleAble = false;
        _this.SelectTargetList = new Array();
        return _this;
    }
    /**
     * @return {?}
     */
    BackWardRule.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new BackWardRule(this.Model);
    };
    /**
     * @protected
     * @return {?}
     */
    BackWardRule.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    BackWardRule.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "TargetRef", this.TargetRef);
        BpmnModelHelper.SetAttributeValue(obj, "SelectTargetAble", this.SelectTargetAble);
        BpmnModelHelper.SetAttributeValue(obj, "WaitReturn", this.WaitReturn);
        BpmnModelHelper.SetAttributeValue(obj, "ResubmitEffect", this.ResubmitEffect);
        BpmnModelHelper.SetAttributeValue(obj, "DynamicBackWardRuleAble", this.DynamicBackWardRuleAble);
        BpmnModelHelper.SetAttributeValue(obj, "SelectTargetList", this.SelectTargetList);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BackWardRule.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.TargetRef = obj.TargetRef;
        this.SelectTargetAble = obj.SelectTargetAble;
        this.WaitReturn = obj.WaitReturn;
        this.ResubmitEffect = obj.ResubmitEffect;
        this.DynamicBackWardRuleAble = obj.DynamicBackWardRuleAble;
        this.SelectTargetList = obj.SelectTargetList;
    };
    /**
     * @return {?}
     */
    BackWardRule.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var backWardRule = new BackWardRule(this.Model);
        backWardRule.TargetRef = this.TargetRef;
        backWardRule.SelectTargetAble = this.SelectTargetAble;
        backWardRule.WaitReturn = this.WaitReturn;
        backWardRule.ResubmitEffect = this.ResubmitEffect;
        backWardRule.DynamicBackWardRuleAble = this.DynamicBackWardRuleAble;
        backWardRule.SelectTargetList = this.SelectTargetList;
        return backWardRule;
    };
    BackWardRule.Clr_Type_ID = "wf.BackWardRule";
    return BackWardRule;
}(ExtensionElement));
export { BackWardRule };
if (false) {
    /** @type {?} */
    BackWardRule.Clr_Type_ID;
    /** @type {?} */
    BackWardRule.prototype.ClrTypeID;
    /** @type {?} */
    BackWardRule.prototype.TargetRef;
    /** @type {?} */
    BackWardRule.prototype.SelectTargetAble;
    /** @type {?} */
    BackWardRule.prototype.WaitReturn;
    /** @type {?} */
    BackWardRule.prototype.ResubmitEffect;
    /** @type {?} */
    BackWardRule.prototype.DynamicBackWardRuleAble;
    /** @type {?} */
    BackWardRule.prototype.SelectTargetList;
}
BpmnModelHelper.RegisterElementType(new BackWardRule(new BpmnModel()));
