// @gsp-wf/wf-process-model/esm5/lib/model/extension/completion-strategy/CounterSignRule.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { Condition } from "../condition-expression/Condition";
import { ExtensionCompletionConditionRuler } from "@edp-pmf/bpmn-model";
var CounterSignRule = /** @class */ (function (_super) {
    tslib_1.__extends(CounterSignRule, _super);
    function CounterSignRule(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = CounterSignRule.Clr_Type_ID;
        return _this;
    }
    /**
     * @return {?}
     */
    CounterSignRule.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new CounterSignRule(this.Model);
    };
    Object.defineProperty(CounterSignRule.prototype, "CounterVariableName", {
        /**
         * 计数器
         * */
        get: /**
         * 计数器
         *
         * @return {?}
         */
        function () {
            return "nrOf" + this.ConclusionValue;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    CounterSignRule.prototype.ResetBeforeLoadFrom = /**
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    CounterSignRule.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "ConclusionValue", this.ConclusionValue);
        BpmnModelHelper.SetAttributeValue(obj, "CalculateType", this.CalculateType);
        BpmnModelHelper.SetAttributeValue(obj, "ConditionValue", this.ConditionValue);
        BpmnModelHelper.SetAttributeValue(obj, "WaitAllVotes", this.WaitAllVotes);
        if (this.CompletionCondition) {
            BpmnModelHelper.SetAttributeValue(obj, "CompletionCondition", this.CompletionCondition.ConvertToJsonObject());
        }
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CounterSignRule.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.ConclusionValue = obj.ConclusionValue;
        this.CalculateType = obj.CalculateType;
        this.ConditionValue = obj.ConditionValue;
        this.WaitAllVotes = obj.WaitAllVotes;
        if (obj.CompletionCondition) {
            /** @type {?} */
            var condition = new Condition(this.Model);
            condition.LoadFromJsonObject(obj.CompletionCondition);
            this.CompletionCondition = condition;
        }
    };
    /**
     * @return {?}
     */
    CounterSignRule.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var counterSignRule = new CounterSignRule(this.Model);
        counterSignRule.ConclusionValue = this.ConclusionValue;
        counterSignRule.CalculateType = this.CalculateType;
        counterSignRule.ConditionValue = this.ConditionValue;
        counterSignRule.WaitAllVotes = this.WaitAllVotes;
        if (this.CompletionCondition) {
            counterSignRule.CompletionCondition = this.CompletionCondition.clone();
        }
        return counterSignRule;
    };
    CounterSignRule.Clr_Type_ID = 'wf.CounterSignRule';
    return CounterSignRule;
}(ExtensionCompletionConditionRuler));
export { CounterSignRule };
if (false) {
    /** @type {?} */
    CounterSignRule.Clr_Type_ID;
    /** @type {?} */
    CounterSignRule.prototype.ClrTypeID;
    /**
     * 结论值
     *
     * @type {?}
     */
    CounterSignRule.prototype.ConclusionValue;
    /**
     * 完成条件
     *
     * @type {?}
     */
    CounterSignRule.prototype.CompletionCondition;
    /**
     * 计算方式
     * @type {?}
     */
    CounterSignRule.prototype.CalculateType;
    /**
     * 满足所有条件值
     * @type {?}
     */
    CounterSignRule.prototype.ConditionValue;
    /**
     * 是否等待所有选票
     * @type {?}
     */
    CounterSignRule.prototype.WaitAllVotes;
}
BpmnModelHelper.RegisterElementType(new CounterSignRule(new BpmnModel()));
