// @gsp-wf/wf-process-model/esm5/lib/model/extension/completion-strategy/ParticipantCoefficient.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelHelper, MexElement } from "@edp-pmf/bpmn-model";
var ParticipantCoefficient = /** @class */ (function (_super) {
    tslib_1.__extends(ParticipantCoefficient, _super);
    function ParticipantCoefficient(model, id, name, code, type, mustCompleted, coefficient) {
        var _this = _super.call(this, model) || this;
        _this.participantId = id || '';
        _this.participantName = name || '';
        _this.participantCode = code || '';
        _this.participantType = type || '';
        _this.mustCompleted = mustCompleted || true;
        _this.coefficient = coefficient || 0;
        return _this;
    }
    Object.defineProperty(ParticipantCoefficient.prototype, "ParticipantId", {
        //序列化内容
        get: 
        //序列化内容
        /**
         * @return {?}
         */
        function () { return this.participantId; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this.participantId = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ParticipantCoefficient.prototype, "ParticipantName", {
        get: /**
         * @return {?}
         */
        function () { return this.participantName; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this.participantName = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ParticipantCoefficient.prototype, "ParticipantCode", {
        get: /**
         * @return {?}
         */
        function () { return this.participantCode; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this.participantCode = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ParticipantCoefficient.prototype, "ParticipantType", {
        get: /**
         * @return {?}
         */
        function () { return this.participantType; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this.participantType = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ParticipantCoefficient.prototype, "MustCompleted", {
        get: /**
         * @return {?}
         */
        function () { return this.mustCompleted; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this.mustCompleted = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ParticipantCoefficient.prototype, "Coefficient", {
        get: /**
         * @return {?}
         */
        function () { return this.coefficient; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this.coefficient = value; },
        enumerable: true,
        configurable: true
    });
    /**
     * @protected
     * @return {?}
     */
    ParticipantCoefficient.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    ParticipantCoefficient.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, 'ParticipantId', this.ParticipantId);
        BpmnModelHelper.SetAttributeValue(obj, 'ParticipantName', this.ParticipantName);
        BpmnModelHelper.SetAttributeValue(obj, 'ParticipantCode', this.ParticipantCode);
        BpmnModelHelper.SetAttributeValue(obj, 'ParticipantType', this.ParticipantType);
        BpmnModelHelper.SetAttributeValue(obj, 'MustCompleted', this.MustCompleted);
        BpmnModelHelper.SetAttributeValue(obj, 'Coefficient', this.Coefficient);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ParticipantCoefficient.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.ParticipantId = obj.ParticipantId;
        this.ParticipantCode = obj.ParticipantCode;
        this.ParticipantName = obj.ParticipantName;
        this.ParticipantType = obj.ParticipantType;
        this.MustCompleted = obj.MustCompleted;
        this.Coefficient = obj.Coefficient;
    };
    return ParticipantCoefficient;
}(MexElement));
export { ParticipantCoefficient };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ParticipantCoefficient.prototype.participantId;
    /**
     * @type {?}
     * @private
     */
    ParticipantCoefficient.prototype.participantName;
    /**
     * @type {?}
     * @private
     */
    ParticipantCoefficient.prototype.participantCode;
    /**
     * @type {?}
     * @private
     */
    ParticipantCoefficient.prototype.participantType;
    /**
     * @type {?}
     * @private
     */
    ParticipantCoefficient.prototype.mustCompleted;
    /**
     * @type {?}
     * @private
     */
    ParticipantCoefficient.prototype.coefficient;
}
