// @farris/ui-section/esm2015/lib/directive/section-toolbar.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, TemplateRef, Input } from '@angular/core';
export class FarrisSectionToolbarDirective {
    /**
     * @param {?} templateRef
     */
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
}
FarrisSectionToolbarDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farrisSectionToolbar]',
            },] }
];
/** @nocollapse */
FarrisSectionToolbarDirective.ctorParameters = () => [
    { type: TemplateRef }
];
FarrisSectionToolbarDirective.propDecorators = {
    toolbarCls: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FarrisSectionToolbarDirective.prototype.toolbarCls;
    /** @type {?} */
    FarrisSectionToolbarDirective.prototype.templateRef;
}
export class FarrisSectionViewChangeDirective {
    /**
     * @param {?} templateRef
     */
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
}
FarrisSectionViewChangeDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farrisSectionViewChange]',
            },] }
];
/** @nocollapse */
FarrisSectionViewChangeDirective.ctorParameters = () => [
    { type: TemplateRef }
];
FarrisSectionViewChangeDirective.propDecorators = {
    viewChangeCls: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FarrisSectionViewChangeDirective.prototype.viewChangeCls;
    /** @type {?} */
    FarrisSectionViewChangeDirective.prototype.templateRef;
}
