// @gsp-wf/rtdevkit/esm2015/lib/runtime/components/task-signadd/task-signadd.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { UserHelperComponent } from '../user-helper/user-helper.component';
import { TaskService } from '../../services/task.service';
import { UserService } from '../../services/user.service';
import { NotifyService } from '@farris/ui-notify';
import { I18nService } from '../../services/i18n.service';
export class TaskSignaddComponent extends UserHelperComponent {
    /**
     * @param {?} taskSvc
     * @param {?} userSvc
     * @param {?} notify
     * @param {?} i18nService
     */
    constructor(taskSvc, userSvc, notify, i18nService) {
        super(i18nService, userSvc);
        this.taskSvc = taskSvc;
        this.userSvc = userSvc;
        this.notify = notify;
        this.i18nService = i18nService;
        this.completed = new EventEmitter();
        this.paginationOptions.id = 'WF-SignAdd-DataTable-Pagination';
    }
    /**
     * @param {?} context
     * @return {?}
     */
    set addsign_context(context) {
        this.selectedUsers = [];
        this.addSignContext = Object.assign({}, context, { assignees: [] });
        this.showModal(!!this.addSignContext.workItemId);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    handleOk() {
        this.selectedUsers = this.userTable.selections;
        // 征询
        if (this.addSignContext.action === 'AddSignFront') {
            // 需要校验选中用户
            if (!this.selectedUsers || this.selectedUsers.length < 1) {
                this.notify.info(this.i18nService.getValue('static.addsign.selectConsultant'));
                return;
            }
            this.addSignContext.assignees = this.selectedUsers.map((/**
             * @param {?} user
             * @return {?}
             */
            user => user.ID));
            this.taskSvc.addSign(this.addSignContext).subscribe((/**
             * @return {?}
             */
            () => {
                this.notify.success(this.i18nService.getValue('static.addsign.successConsult'));
                this.completed.emit(true);
                this.showModal(false);
            }), (/**
             * @param {?} error
             * @return {?}
             */
            error => {
                if (error && error.error && error.error.Message) {
                    this.notify.error(error.error.Message);
                }
            }));
        }
        else {
            // 加签
            this.completed.emit(true);
            this.showModal(false);
        }
    }
}
TaskSignaddComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf-task-signadd',
                template: "<farris-dialog #allUserDialog [buttons]=\"buttonRef\" [width]=\"850\" [height]=\"520\" [title]=\"'\u7CFB\u7EDF\u7528\u6237'\">\r\n  <div class=\"d-flex flex-column my-data-grid h-100\" style=\"padding: 0 14px;\">\r\n    <div class=\"flex-auto\" style=\"position:relative;\">\r\n      <farris-datatable #allUserTable [columns]=\"columns\" [data]=\"userListData | paginate: paginationOptions\"\r\n        [showFilterBar]='true' [singleSelect]=\"singleSelect\" [idField]=\"'ID'\" [remote]=\"'server'\" (search)=\"search($event)\"\r\n        [pagination]='false'></farris-datatable>\r\n    </div>\r\n    <div style=\"position:relative;\">\r\n      <pagination-controls [id]=\"paginationOptions.id\" [previousLabel]=\"' '\" [nextLabel]=\"' '\"\r\n        (pageChange)=\"onPageChange($event)\" (pageSizeChange)=\"onPageSizeChange($event)\">\r\n      </pagination-controls>\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n<ng-template #buttonRef>\r\n  <button class=\"btn btn-sm btn-primary px-3\" (click)=\"handleOk()\">\u786E\u5B9A</button>\r\n  <button class=\"btn btn-sm btn-default px-3\" (click)=\"handleCancel()\">\u53D6\u6D88</button>\r\n</ng-template>",
                styles: [".my-data-grid .filter-row .search-fields{padding-left:0!important}.my-data-grid .filter-row .search-text{padding-right:0!important}"]
            }] }
];
/** @nocollapse */
TaskSignaddComponent.ctorParameters = () => [
    { type: TaskService },
    { type: UserService },
    { type: NotifyService },
    { type: I18nService }
];
TaskSignaddComponent.propDecorators = {
    addsign_context: [{ type: Input }],
    completed: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    TaskSignaddComponent.prototype.completed;
    /** @type {?} */
    TaskSignaddComponent.prototype.addSignContext;
    /**
     * @type {?}
     * @private
     */
    TaskSignaddComponent.prototype.taskSvc;
    /**
     * @type {?}
     * @protected
     */
    TaskSignaddComponent.prototype.userSvc;
    /**
     * @type {?}
     * @protected
     */
    TaskSignaddComponent.prototype.notify;
    /**
     * @type {?}
     * @protected
     */
    TaskSignaddComponent.prototype.i18nService;
}
