// @gsp-wf/wf-dimension-def/esm5/lib/service/wf-dimension-def.viewModel.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpService } from '@ecp-caf/caf-common';
import { Injectable } from '@angular/core';
var WFDimensionDefViewmodel = /** @class */ (function () {
    function WFDimensionDefViewmodel(http) {
        this.http = http;
        this.url = '/api/runtime/wf/v1.0/dimensionDef';
    }
    /**
     * @param {?} flowFormId
     * @return {?}
     */
    WFDimensionDefViewmodel.prototype.loadDimensionDefs = /**
     * @param {?} flowFormId
     * @return {?}
     */
    function (flowFormId) {
        if (flowFormId) {
            /** @type {?} */
            var url = this.url + ("?entryFormID=" + flowFormId);
            return this.http.get(url);
        }
    };
    /**
     * @param {?} id
     * @return {?}
     */
    WFDimensionDefViewmodel.prototype.removeDimensionDef = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        if (id) {
            /** @type {?} */
            var url = this.url + ("/" + id);
            return this.http.delete(url, {});
        }
    };
    /**
     * @param {?} dimensionDef
     * @return {?}
     */
    WFDimensionDefViewmodel.prototype.addDimensionDef = /**
     * @param {?} dimensionDef
     * @return {?}
     */
    function (dimensionDef) {
        return this.http.post(this.url, dimensionDef);
    };
    /**
     * @param {?} dimensionDef
     * @return {?}
     */
    WFDimensionDefViewmodel.prototype.updateDimensionDef = /**
     * @param {?} dimensionDef
     * @return {?}
     */
    function (dimensionDef) {
        return this.http.patch(this.url, dimensionDef);
    };
    /**
     * @param {?} flowFormId
     * @return {?}
     */
    WFDimensionDefViewmodel.prototype.getDimensionDefSet = /**
     * @param {?} flowFormId
     * @return {?}
     */
    function (flowFormId) {
        /** @type {?} */
        var url = "/api/runtime/wf/v1.0/dimensionDef/customSetting?entryFormId=" + flowFormId;
        return this.http.get(url);
    };
    WFDimensionDefViewmodel.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    WFDimensionDefViewmodel.ctorParameters = function () { return [
        { type: HttpService }
    ]; };
    return WFDimensionDefViewmodel;
}());
export { WFDimensionDefViewmodel };
if (false) {
    /**
     * @type {?}
     * @private
     */
    WFDimensionDefViewmodel.prototype.url;
    /**
     * @type {?}
     * @private
     */
    WFDimensionDefViewmodel.prototype.http;
}
