// @farris/ui-html-editor/esm2015/lib/image-drop-module/image-drop.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class ImageDrop {
    /**
     * @param {?} quill
     * @param {?=} options
     * @param {?=} ngZone
     */
    constructor(quill, options = {}, ngZone) {
        this.ngZone = ngZone;
        // save the quill reference
        this.quill = quill;
        // bind handlers to this instance
        this.handleDrop = this.handleDrop.bind(this);
        this.handlePaste = this.handlePaste.bind(this);
        // listen for drop and paste events
        this.quill.root.addEventListener('drop', this.handleDrop, false);
        this.quill.root.addEventListener('paste', this.handlePaste, false);
    }
    /**
     * @param {?} evt
     * @return {?}
     */
    handleDrop(evt) {
        evt.preventDefault();
        if (evt.dataTransfer &&
            evt.dataTransfer.files &&
            evt.dataTransfer.files.length) {
            if (document.caretRangeFromPoint) {
                /** @type {?} */
                const selection = document.getSelection();
                /** @type {?} */
                const range = document.caretRangeFromPoint(evt.clientX, evt.clientY);
                if (selection && range) {
                    selection.setBaseAndExtent(range.startContainer, range.startOffset, range.startContainer, range.startOffset);
                }
            }
            this.readFiles(evt.dataTransfer.files, this.insert.bind(this));
        }
    }
    /**
     * @param {?} evt
     * @return {?}
     */
    handlePaste(evt) {
        if (evt) {
            evt.preventDefault();
        }
        if (evt.clipboardData &&
            evt.clipboardData.items &&
            evt.clipboardData.items.length) {
            this.readFiles(evt.clipboardData.items, (/**
             * @param {?} dataUrl
             * @return {?}
             */
            dataUrl => {
                // const selection = this.quill.getSelection();
                // if (selection && selection.length) {
                //     // we must be in a browser that supports pasting (like Firefox)
                //     // so it has already been placed into the editor
                // } else {
                //     // otherwise we wait until after the paste when this.quill.getSelection()
                //     // will return a valid index
                // const selection = this.quill.getSelection();
                // if (selection && selection.length) {
                //     // we must be in a browser that supports pasting (like Firefox)
                //     // so it has already been placed into the editor
                // } else {
                //     // otherwise we wait until after the paste when this.quill.getSelection()
                //     // will return a valid index
                //     let _ngZone = this.ngZone;
                //     if (!_ngZone && this.quill && this.quill.options) {
                //         _ngZone = this.quill.options.ngZone;
                //     }
                //     if (_ngZone) {
                //         _ngZone.runOutsideAngular(() => {
                //             setTimeout(() => this.insert(dataUrl), 0);
                //         });
                //     }
                // }
                /** @type {?} */
                let _ngZone = this.ngZone;
                if (!_ngZone && this.quill && this.quill.options) {
                    _ngZone = this.quill.options.ngZone;
                }
                if (_ngZone) {
                    _ngZone.runOutsideAngular((/**
                     * @return {?}
                     */
                    () => {
                        setTimeout((/**
                         * @return {?}
                         */
                        () => {
                            /** @type {?} */
                            const img = this.quill.container.querySelector('img[src="' + dataUrl + '"]');
                            if (!img) {
                                this.insert(dataUrl);
                            }
                        }), 0);
                    }));
                }
            }));
        }
    }
    /**
     * @param {?} dataUrl
     * @return {?}
     */
    insert(dataUrl) {
        /** @type {?} */
        const index = (this.quill.getSelection() || {}).index || this.quill.getLength();
        this.quill.insertEmbed(index, 'image', dataUrl, 'user');
    }
    /**
     * @param {?} files
     * @param {?} callback
     * @return {?}
     */
    readFiles(files, callback) {
        // check each file for an image
        /** @type {?} */
        let imgs = [];
        const { limitImageSize, maxImgSize } = this.quill.options;
        /** @type {?} */
        var uploadImgs = Array.from(files).filter((/**
         * @param {?} n
         * @return {?}
         */
        (n) => {
            /** @type {?} */
            const blob = n.getAsFile && n.getAsFile() || n;
            if (!blob.type.match(/^image\/(gif|jpe?g|a?png|svg|webp|bmp|vnd\.microsoft\.icon)/i)) {
                // file is not an image
                // Note that some file formats such as psd start with image/* but are not readable
                return false;
            }
            else {
                if (!limitImageSize) {
                    return true;
                }
                else {
                    if (maxImgSize && blob.size > maxImgSize * 1024) {
                        imgs.push(blob);
                        return false;
                    }
                    return true;
                }
            }
        }));
        uploadImgs.forEach((/**
         * @param {?} file
         * @return {?}
         */
        (file) => {
            /** @type {?} */
            const reader = new FileReader();
            reader.onload = (/**
             * @param {?} evt
             * @return {?}
             */
            (evt) => {
                callback(evt.target.result);
            });
            /** @type {?} */
            const blob = file.getAsFile && file.getAsFile() || file;
            if (blob instanceof Blob) {
                reader.readAsDataURL(blob);
            }
        }));
        // [].forEach.call(files, file => {
        //     // read the clipboard item or file
        //     const blob = file.getAsFile && file.getAsFile() || file;
        //     if (!blob.type.match(/^image\/(gif|jpe?g|a?png|svg|webp|bmp|vnd\.microsoft\.icon)/i)) {
        //         // file is not an image
        //         // Note that some file formats such as psd start with image/* but are not readable
        //         return;
        //     }
        //     if (limitImageSize && maxImgSize && blob.size > maxImgSize * 1024) {
        //         imgs.push(blob);
        //         return;
        //     } else {
        //         // set up file reader
        //         const reader = new FileReader();
        //         reader.onload = (evt: any) => {
        //             callback(evt.target.result);
        //         };
        //         if (blob instanceof Blob) {
        //             reader.readAsDataURL(blob);
        //         }
        //     }
        // });
        if (imgs && imgs.length) {
            this.quill.options.showMaxImageMessage(imgs);
        }
    }
}
if (false) {
    /** @type {?} */
    ImageDrop.prototype.quill;
    /**
     * @type {?}
     * @private
     */
    ImageDrop.prototype.ngZone;
}
