// @farris/devkit/esm5/lib/misc/router-param.service.js
import { Injectable } from '@angular/core';
/**
 * 路由参数服务
 */
var RouterParamService = /** @class */ (function () {
    function RouterParamService() {
    }
    /**
     * 获取参数
     */
    RouterParamService.prototype.getParams = function (url) {
        var allParams = this.getAllParams();
        return allParams[url] || {};
    };
    /**
     * 设置参数
     */
    RouterParamService.prototype.setParams = function (url, params) {
        var allParams = this.getAllParams();
        allParams[url] = params;
        this.setAllParams(allParams);
    };
    /**
     * 清空参数
     */
    RouterParamService.prototype.clearParams = function () {
        throw new Error('Not Implemented');
    };
    /**
     * 获取全部参数
     */
    RouterParamService.prototype.getAllParams = function () {
        var allParamsJson = window.sessionStorage.getItem('ROUTER_PARAMS') || '{}';
        var allParams = JSON.parse(allParamsJson);
        return allParams;
    };
    /**
     * 设置所有的参数
     */
    RouterParamService.prototype.setAllParams = function (allParams) {
        allParams = allParams || {};
        var allParamsJson = JSON.stringify(allParams);
        window.sessionStorage.setItem('ROUTER_PARAMS', allParamsJson);
    };
    /**
     * 清空所有参数
     */
    RouterParamService.prototype.clearAllParams = function () {
        window.sessionStorage.setItem('ROUTER_PARAMS', '{}');
    };
    RouterParamService.decorators = [
        { type: Injectable }
    ];
    return RouterParamService;
}());
export { RouterParamService };
