// @gsp-sys/rac-common/esm2015/lib/component/auobj/auobj-help/auobj-help.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Output, Input, EventEmitter } from '@angular/core';
import { from } from 'rxjs';
import { LookupComponent } from '@farris/ui-lookup';
import { AuthobjEntity } from '../../../common/entity/authobj.entity';
export class AuobjHelpComponent {
    constructor() {
        this.displayText = '';
        this.selectedKeyOld = '';
        /**
         * 导航事件
         */
        this.selectedChanged = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.curKeyIdList = [];
        this.curKeyIdList.push(this.curKeyId);
    }
    // 取数列
    /**
     * @param {?} args
     * @return {?}
     */
    afterKeySelected(args) {
        this.curAuthObj = new AuthobjEntity();
        this.curKeyIdList = [];
        // 更新值
        from(args).subscribe((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            this.curAuthObj.id = item['id'];
            this.curAuthObj.name = item['name'];
            this.curKeyId = item['id'];
            this.curKeyIdList.push(item['id']);
            this.displayText = item['name'];
            this.selectedChanged.emit(this.curAuthObj);
        }));
        this.auobjLookup.closeDialog();
    }
    /**
     * 选中行改变后触发事件
     * @param {?} args
     * @return {?}
     */
    handleSelection(args) {
        this.curKeyId = args.node.data.id;
        this.curAuthObj = args.node.data;
        // 首次触发
        if (this.curKeyId === '') {
            this.selectedKeyOld = this.curKeyId;
            this.selectedChanged.emit(this.curAuthObj);
        }
        else {
            // 变化的情况下触发
            if (this.selectedKeyOld !== this.curKeyId) {
                this.selectedKeyOld = this.curKeyId;
                this.selectedChanged.emit(this.curAuthObj);
            }
        }
    }
}
AuobjHelpComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'rac-auobj-help',
                template: "<farris-lookup #auobjLookup [displayText]=\"displayText\"  name=\"auobjLookup\"\r\n[showButtons]=\"true\" [title]=\"'\u8BF7\u9009\u62E9\u6743\u9650\u5BF9\u8C61'\" [buttonsRef]=\"this.auobjHelp.btnOkAndCancleRef\">\r\n  <div style=\"padding: 10px\">\r\n    <rac-auobj-tree #auobjHelp [initSelectedKeys]=\"this.curKeyIdList\" [dataMode] = 'dataMode'\r\n      (selectedAfterOk)=\"afterKeySelected($event)\" (closeModalEmitter)=\"this.auobjLookup.closeDialog()\">\r\n    </rac-auobj-tree>\r\n  </div>\r\n</farris-lookup>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
AuobjHelpComponent.ctorParameters = () => [];
AuobjHelpComponent.propDecorators = {
    dataMode: [{ type: Input }],
    displayText: [{ type: Input }],
    curKeyId: [{ type: Input }],
    auobjLookup: [{ type: ViewChild, args: ['auobjLookup',] }],
    selectedChanged: [{ type: Output }]
};
if (false) {
    /**
     * 要加载的数据类型
     * all代表全部权限对象
     * reassign代表允许在用户上重新授权的权限对象
     * @type {?}
     */
    AuobjHelpComponent.prototype.dataMode;
    /** @type {?} */
    AuobjHelpComponent.prototype.displayText;
    /** @type {?} */
    AuobjHelpComponent.prototype.curKeyId;
    /** @type {?} */
    AuobjHelpComponent.prototype.curKeyIdList;
    /** @type {?} */
    AuobjHelpComponent.prototype.curAuthObj;
    /** @type {?} */
    AuobjHelpComponent.prototype.selectedKeyOld;
    /** @type {?} */
    AuobjHelpComponent.prototype.auobjLookup;
    /**
     * 导航事件
     * @type {?}
     */
    AuobjHelpComponent.prototype.selectedChanged;
}
