// @gsp-aif/runtime-api/esm5/lib/aif-documents-track/aif-documents-track-manager.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { ButtonsModule } from '@progress/kendo-angular-buttons';
import { GridModule } from '@progress/kendo-angular-grid';
import { WindowModule, DialogModule } from '@progress/kendo-angular-dialog';
import { DropDownsModule } from '@progress/kendo-angular-dropdowns';
import { TreeViewModule } from '@progress/kendo-angular-treeview';
import { SplitterModule } from '@progress/kendo-angular-layout';
import { AifDocumentsTrackService } from './service/aif-documents-track.service';
import { HttpService, SessionService } from '@ecp-caf/caf-common';
import { AifDocumentsTrackManagerComponent } from './aif-documents-track-manager/aif-documents-track-manager.component';
import { AifDocumentsTreeComponent } from './aif-documents-tree/aif-documents-tree.component';
import { ModalModule } from '@farris/ui-modal';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { AifDocumentsTrackManagerModuleConfig } from '../entity/aif-documents-track-manager-module-config';
var ɵ0 = {};
var AifDocumentsTrackManagerModule = /** @class */ (function () {
    function AifDocumentsTrackManagerModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    AifDocumentsTrackManagerModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            "ngModule": AifDocumentsTrackManagerModule,
            "providers": [
                { provide: AifDocumentsTrackManagerModuleConfig, useValue: config || {} }
            ]
        };
    };
    AifDocumentsTrackManagerModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        HttpClientModule,
                        FormsModule,
                        ReactiveFormsModule,
                        ButtonsModule,
                        GridModule,
                        WindowModule,
                        DropDownsModule,
                        DialogModule,
                        TreeViewModule,
                        SplitterModule,
                        GridModule,
                        ModalModule.forRoot(),
                        FarrisDialogModule,
                    ],
                    providers: [
                        SessionService,
                        HttpService,
                        { provide: AifDocumentsTrackManagerModuleConfig, useValue: ɵ0 },
                        AifDocumentsTrackService,
                    ],
                    declarations: [
                        AifDocumentsTrackManagerComponent,
                        AifDocumentsTreeComponent,
                    ],
                    bootstrap: [
                        AifDocumentsTrackManagerComponent
                    ],
                    entryComponents: [
                        AifDocumentsTrackManagerComponent
                    ],
                    exports: [
                        AifDocumentsTrackManagerComponent
                    ],
                },] }
    ];
    return AifDocumentsTrackManagerModule;
}());
export { AifDocumentsTrackManagerModule };
export { ɵ0 };
