// @farris/mobile-command-services/esm5/lib/js-bridge-service/ccwork-service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/js-bridge-service/ccwork-service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Subject } from 'rxjs';
var CCworkService = /** @class */ (function () {
    function CCworkService() {
    }
    /**
     * @param {?} callback
     * @return {?}
     */
    CCworkService.prototype.init = /**
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        window.top['iCityBridge'].init(callback);
    };
    /**
     * {
     * quality:图片质量，可选值high、mid、low, high为1920*1080；mid为720*540；low: 480*360，默认为mid
     * mode:唤起模式，可选值both（仅支持iOS）、camera、album, 默认为camera;
     * mark文字水印，字数最大值50
     * }
     * @param options Object
     *
     */
    /**
     * {
     * quality:图片质量，可选值high、mid、low, high为1920*1080；mid为720*540；low: 480*360，默认为mid
     * mode:唤起模式，可选值both（仅支持iOS）、camera、album, 默认为camera;
     * mark文字水印，字数最大值50
     * }
     * @param {?=} options Object
     *
     * @return {?}
     */
    CCworkService.prototype.takePhoto = /**
     * {
     * quality:图片质量，可选值high、mid、low, high为1920*1080；mid为720*540；low: 480*360，默认为mid
     * mode:唤起模式，可选值both（仅支持iOS）、camera、album, 默认为camera;
     * mark文字水印，字数最大值50
     * }
     * @param {?=} options Object
     *
     * @return {?}
     */
    function (options) {
        if (options === void 0) { options = { quality: 'mid' }; }
        // callback return object {errormessage:""，result:"base64格式"}
        /** @type {?} */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                subject.next(data['result']);
            }
        });
        window.top['iCityBridge'].ccworkGetPicsFromCameraAndAlbum(tslib_1.__assign({}, options), callback);
        return subject;
    };
    /**
     * path 如果不为空则从指定图片文件识别二维码,空就吊起相机
     * filters 条码类型过滤列表，如QR、EAN13、EAN8
     *
     * @param options Object
     *
     */
    /**
     * path 如果不为空则从指定图片文件识别二维码,空就吊起相机
     * filters 条码类型过滤列表，如QR、EAN13、EAN8
     *
     * @return {?}
     */
    CCworkService.prototype.scanQRCode = /**
     * path 如果不为空则从指定图片文件识别二维码,空就吊起相机
     * filters 条码类型过滤列表，如QR、EAN13、EAN8
     *
     * @return {?}
     */
    function () {
        /** @type {?} */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                subject.next(data['result']);
            }
        });
        window.top['iCityBridge'].ccworkScanCode({}, callback);
        return subject;
    };
    /**
     * @return {?}
     */
    CCworkService.prototype.getSysInfo = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    CCworkService.prototype.getLocation = /**
     * @return {?}
     */
    function () {
        /**
         *  回调函数
         *  输出结构 object
         *  errormessage:""
         *  result:Object {
         *  addr:'山东省济南市历下区经十路靠近浪潮'
         *  city:"济南"
         *  country:"中国"
         *  course:-1
         *  district:"历下区"
         *  lat:36.66254123263889
         *  lng:117.12974745008681
         *  province:"山东省"
         *  street:"经十路"
         *  streetNum:"1单元"
         * }
         *  status:true
         * @type {?}
         */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                subject.next(data['result']);
            }
        });
        window.top['iCityBridge'].ccworkGetLocation(callback);
        return subject;
    };
    // TODO:云上协同选择文件
    // TODO:云上协同选择文件
    /**
     * @return {?}
     */
    CCworkService.prototype.selectFile = 
    // TODO:云上协同选择文件
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var subject = new Subject();
        subject.next([]);
        return subject;
    };
    /**
     *
     * @param number string
     * @param content string
     */
    /**
     *
     * @param {?} number string
     * @param {?} content string
     * @return {?}
     */
    CCworkService.prototype.sendSMS = /**
     *
     * @param {?} number string
     * @param {?} content string
     * @return {?}
     */
    function (number, content) {
        /** @type {?} */
        var options = {
            numbers: [number],
            content: content
        };
        window.top['iCityBridge'].ccworkSendSms(options);
    };
    /**
     * @param {?} number
     * @param {?} content
     * @return {?}
     */
    CCworkService.prototype.openSendSMS = /**
     * @param {?} number
     * @param {?} content
     * @return {?}
     */
    function (number, content) {
        window.top['iCityBridge'].sendSMS({ tel: number, msg: content });
    };
    /**
     *
     * @param number string
     */
    /**
     *
     * @param {?} number string
     * @return {?}
     */
    CCworkService.prototype.tel = /**
     *
     * @param {?} number string
     * @return {?}
     */
    function (number) {
        window.top['iCityBridge'].ringUp(number);
    };
    /**
     *
     * @param options
     */
    /**
     *
     * @param {?=} options
     * @return {?}
     */
    CCworkService.prototype.selectPictureFromAlbum = /**
     *
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        if (options === void 0) { options = { maxNum: 9, mType: 0 }; }
        /**
         * params object
         *  { maxNum : number 9 最大可选数 默认9
         *    mType: number 0 	选择模式，可选值0/1, 0:图片，1：视频;
         * }
         * @type {?}
         */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                subject.next(data['result']);
            }
        });
        window.top['iCityBridge'].ccworkGetPicsFromAlbum(options, callback);
        return subject;
        /**
         *  object
         *  errormessage:""
         *  result:Array [
         * base64
         * ]
         *  status:true
         */
    };
    /**
     *
     * @param options
     */
    /**
     *
     * @param {?} options
     * @return {?}
     */
    CCworkService.prototype.takeVideo = /**
     *
     * @param {?} options
     * @return {?}
     */
    function (options) {
        /**
         * params object
         *  {
         *  time : number 短视频时间（秒） 最大60s
         *  quality: number 0 	视频质量 取值范围0~1
         *  id :string 	视频id（视频名字）
         * }
         * @type {?}
         */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                subject.next(data['result']);
            }
        });
        window.top['iCityBridge'].ccworkTakeShortVideo(tslib_1.__assign({ time: 15 }, options), callback);
        return subject;
        /**
         *  object
         *  errormessage:""
         *  result:Object {
         *  value:'path'
         * }
         *  status:true
         */
    };
    /**
     * @return {?}
     */
    CCworkService.prototype.closeWindow = /**
     * @return {?}
     */
    function () {
        window.top['iCityBridge'].closeWebView();
    };
    /**
     *
     * @param title string
     */
    /**
     *
     * @param {?} title string
     * @return {?}
     */
    CCworkService.prototype.setTitles = /**
     *
     * @param {?} title string
     * @return {?}
     */
    function (title) {
        window.top['iCityBridge'].setTitle(title);
    };
    /**
     * @param {?} options
     * @return {?}
     */
    CCworkService.prototype.downloadFile = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        alert('暂不支持附件下载');
    };
    /**
     * @param {?=} fn
     * @return {?}
     */
    CCworkService.prototype.setOriginGoback = /**
     * @param {?=} fn
     * @return {?}
     */
    function (fn) {
        window.top['iCityBridge'].ccworkInterceptOriginGoback(!!fn, fn, (/**
         * @return {?}
         */
        function () { }));
    };
    /**
     * @param {?=} fn
     * @return {?}
     */
    CCworkService.prototype.pushOriginGoback = /**
     * @param {?=} fn
     * @return {?}
     */
    function (fn) {
        window['OriginGoback'] = window['OriginGoback'] || [];
        /** @type {?} */
        var OriginGoback = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var gobackFn = window['OriginGoback'].pop();
            if (gobackFn) {
                gobackFn();
            }
            else {
                window.top['iCityBridge'].ccworkInterceptOriginGoback(false, (/**
                 * @return {?}
                 */
                function () { }), (/**
                 * @return {?}
                 */
                function () { }));
                history.back();
            }
        });
        if (fn) {
            window['OriginGoback'].push(fn);
            window.top['iCityBridge'].ccworkInterceptOriginGoback(true, OriginGoback, (/**
             * @return {?}
             */
            function () { }));
        }
    };
    /**
     * @param {?=} fn
     * @return {?}
     */
    CCworkService.prototype.popOriginGoback = /**
     * @param {?=} fn
     * @return {?}
     */
    function (fn) {
        window['OriginGoback'].pop();
    };
    /**
     * @param {?} opt
     * @return {?}
     */
    CCworkService.prototype.playShortVideo = /**
     * @param {?} opt
     * @return {?}
     */
    function (opt) {
        window.top['iCityBridge'].ccworkPlayShortVideo(opt);
    };
    return CCworkService;
}());
export { CCworkService };
