// @farris/mobile-command-services/esm5/lib/js-bridge-service/cloud-plus-service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/js-bridge-service/cloud-plus-service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Subject } from 'rxjs';
var CloudPlusService = /** @class */ (function () {
    function CloudPlusService() {
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    CloudPlusService.prototype.playShortVideo = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        throw new Error('Method not implemented.');
    };
    // 目前拍完返回报错
    // 目前拍完返回报错
    /**
     * @param {?=} options
     * @return {?}
     */
    CloudPlusService.prototype.takePhoto = 
    // 目前拍完返回报错
    /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        if (options === void 0) { options = {
            quality: 75,
            encodingType: 0,
            targetWidth: 600,
            targetHeight: 600
        }; }
        /** @type {?} */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data && data['originalData']) {
                subject.next(data['originalData']);
            }
        });
        /** @type {?} */
        var errorCallback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                console.log('takePhoto', data);
                subject.next([]);
                return;
            }
        });
        window['imp'].iCamera.open(options, callback, errorCallback);
        return subject;
    };
    /**
     * @return {?}
     */
    CloudPlusService.prototype.scanQRCode = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                console.log('二维码回调');
                subject.next(data);
            }
        });
        console.log('调用二维码');
        window['imp'].iBarCode.scan(callback);
        return subject;
    };
    /**
     * @return {?}
     */
    CloudPlusService.prototype.getLocation = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                var latitude = data.latitude, longitude = data.longitude, addr = data.addr, city = data.city, country = data.country, district = data.district, province = data.province, street = data.street, streetNum = data.streetNum;
                subject.next({
                    addr: addr,
                    city: city,
                    country: country,
                    district: district,
                    province: province,
                    street: street,
                    streetNum: streetNum,
                    lat: latitude,
                    lng: longitude,
                });
            }
        });
        window['imp'].iGps.getInfo(callback);
        return subject;
        // iphone 打印输出信息
        /**
         * {latitude:"36.662413",longitude:"117.129769"}
         */
        // 安卓
        /**
         *  回调函数
         *  输出结构 object
         *  errormessage:""
         *  result:Object {
         *  addr:'山东省济南市历下区经十路靠近浪潮'
         *  city:"济南"
         *  country:"中国"
         *  course:-1
         *  district:"历下区"
         *  lat:36.66254123263889
         *  lng:117.12974745008681
         *  province:"山东省"
         *  street:"经十路"
         *  streetNum:"1单元"
         * }
         *  status:true
         */
    };
    /**
     * @return {?}
     */
    CloudPlusService.prototype.getSysInfo = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                subject.next(data);
            }
        });
        window['imp'].iDevice.sysInfo(callback);
        return subject;
    };
    /**
     * @param {?} number
     * @param {?} content
     * @return {?}
     */
    CloudPlusService.prototype.sendSMS = /**
     * @param {?} number
     * @param {?} content
     * @return {?}
     */
    function (number, content) {
        window['imp'].iSms.send(number, content);
    };
    /**
     * @param {?} number
     * @param {?} content
     * @return {?}
     */
    CloudPlusService.prototype.openSendSMS = /**
     * @param {?} number
     * @param {?} content
     * @return {?}
     */
    function (number, content) {
        window['imp'].iSms.open(number, content);
    };
    /**
     * @param {?} number
     * @return {?}
     */
    CloudPlusService.prototype.tel = /**
     * @param {?} number
     * @return {?}
     */
    function (number) {
        window['imp'].iTel.dial(number);
    };
    /**
     * @param {?=} options
     * @return {?}
     */
    CloudPlusService.prototype.selectPictureFromAlbum = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        if (options === void 0) { options = { maxNum: 9, mType: 0 }; }
        /** @type {?} */
        var defaults = {
            num: 9,
            quality: 75,
            destinationType: 0,
            encodingType: 1,
            targetWidth: 600,
            targetHeight: 600
        };
        defaults.num = options.maxNum;
        /** @type {?} */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                subject.next(data.map((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    return item.originalData;
                })));
            }
        });
        /** @type {?} */
        var errorCallback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                console.log('selectPictureFromAlbum', data);
                subject.next([]);
            }
        });
        window['imp'].iCamera.select(tslib_1.__assign({}, defaults, options), callback, errorCallback);
        return subject;
    };
    /**
     * @private
     * @param {?=} options
     * @return {?}
     */
    CloudPlusService.prototype.selectFileFromMobile = /**
     * @private
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        if (options === void 0) { options = { maximum: 6, fileType: '.jpg|.png' }; }
    };
    /**
     * @private
     * @param {?} path
     * @return {?}
     */
    CloudPlusService.prototype.getBase64 = /**
     * @private
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                subject.next(data);
            }
        });
        /** @type {?} */
        var errorCallback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                console.log('getBase64', data);
                subject.next([]);
            }
        });
        window['imp'].iFile.getBase64(path, callback, errorCallback);
        return subject;
    };
    /**
     * @param {?=} options
     * @return {?}
     */
    CloudPlusService.prototype.selectFile = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        if (options === void 0) { options = { maximum: 6 }; }
        /** @type {?} */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data && data.length > 0) {
                /** @type {?} */
                var paths_1 = [];
                /** @type {?} */
                var count_1 = 0;
                data.forEach((/**
                 * @param {?} dataItem
                 * @return {?}
                 */
                function (dataItem) {
                    window['imp'].iFile.getBase64(dataItem.path, (/**
                     * @param {?} bs64
                     * @return {?}
                     */
                    function (bs64) {
                        paths_1.push({
                            name: data[count_1].name,
                            size: data[count_1].size,
                            base64: bs64
                        });
                        count_1++;
                        if (data.length === paths_1.length) {
                            // 防止触发多次next  一次性把所有数据抛出去
                            subject.next(paths_1);
                        }
                    }), (/**
                     * @param {?} errorMsg
                     * @return {?}
                     */
                    function (errorMsg) {
                        if (errorMsg) {
                            console.log('getBase64', errorMsg);
                            subject.next([]);
                        }
                    }));
                }));
            }
        });
        /** @type {?} */
        var errorCallback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                console.log('selectFile', data);
                subject.next([]);
            }
        });
        window['imp'].iFile.select(tslib_1.__assign({}, options), callback, errorCallback);
        return subject;
    };
    /**
     * @param {?} options
     * @return {?}
     */
    CloudPlusService.prototype.takeVideo = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        /** @type {?} */
        var option = { id: options.id };
        if (options.quality) {
            if (options.quality <= 0.5) {
                option['fps'] = 30;
            }
            else if (options.quality <= 1) {
                option['fps'] = 60;
            }
        }
        else {
            option['fps'] = 30;
        }
        /** @type {?} */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                subject.next(data);
            }
        });
        /** @type {?} */
        var errorCallback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                console.log('takeVideo', data);
                subject.next([]);
            }
        });
        window['imp'].iVideo.record(option, callback, errorCallback);
        return subject;
    };
    /**
     * @return {?}
     */
    CloudPlusService.prototype.closeWindow = /**
     * @return {?}
     */
    function () {
        window['imp'].iWindow.close();
    };
    /**
     * @param {?} title
     * @return {?}
     */
    CloudPlusService.prototype.setTitles = /**
     * @param {?} title
     * @return {?}
     */
    function (title) {
        document.title = title;
    };
    /**
     * @param {?} options
     * @return {?}
     */
    CloudPlusService.prototype.downloadFile = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        /** @type {?} */
        var subject = new Subject();
        /** @type {?} */
        var callback = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data.status !== 1) {
                console.log(data);
            }
            subject.next(true);
        });
        window['imp'].iFile.download(options, callback);
        return subject;
    };
    /**
     * @param {?=} fn
     * @return {?}
     */
    CloudPlusService.prototype.setOriginGoback = /**
     * @param {?=} fn
     * @return {?}
     */
    function (fn) {
        /** @type {?} */
        var isAGreaterThanB = (/**
         * @param {?} parameA
         * @param {?} parameB
         * @return {?}
         */
        function (parameA, parameB) {
            /** @type {?} */
            var length = parameA.length > parameB.length ? parameB.length : parameA.length;
            for (var i = 0; i < length; i++) {
                if (parseInt(parameB[i]) > parseInt(parameA[i])) {
                    return false;
                }
            }
            return true;
        });
        /** @type {?} */
        var setFunction = (/**
         * @return {?}
         */
        function () {
            if (fn) {
                window['imp'].iWindow.onBackKeyDown2(fn);
                isAGreaterThanB(currentVersion.split('.'), minVersion.split('.')) && window['imp'].iWindow.onTitleBackKeyDown2(fn);
            }
            else {
                window['imp'].iWindow.cancelBackKeyDown();
                isAGreaterThanB(currentVersion.split('.'), minVersion.split('.')) && window['imp'].iWindow.cancelTitleBackKeyDown();
            }
        });
        /** @type {?} */
        var minVersion = '4.8.21';
        /** @type {?} */
        var currentVersion = window['CURRENT_VERSION'];
        if (currentVersion) {
            setFunction();
        }
        else {
            window['imp'].iDevice.sysInfo((/**
             * @param {?} info
             * @return {?}
             */
            function (info) {
                /** @type {?} */
                var newVersion = info.appVersion.split('-');
                window['CURRENT_VERSION'] = newVersion[0];
                currentVersion = window['CURRENT_VERSION'];
                setFunction();
            }));
        }
    };
    /**
     * @param {?=} fn
     * @return {?}
     */
    CloudPlusService.prototype.pushOriginGoback = /**
     * @param {?=} fn
     * @return {?}
     */
    function (fn) {
        window['OriginGoback'] = window['OriginGoback'] || [];
        /** @type {?} */
        var OriginGoback = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var gobackFn = window['OriginGoback'].pop();
            if (gobackFn) {
                gobackFn();
            }
            else {
                window['imp'].iWindow.cancelBackKeyDown();
                window['imp'].iWindow.cancelTitleBackKeyDown();
                history.back();
            }
        });
        if (fn) {
            window['OriginGoback'].push(fn);
            window['imp'].iWindow.onBackKeyDown2(OriginGoback);
            window['imp'].iWindow.onTitleBackKeyDown2(OriginGoback);
        }
    };
    /**
     * @return {?}
     */
    CloudPlusService.prototype.popOriginGoback = /**
     * @return {?}
     */
    function () {
        window['OriginGoback'].pop();
    };
    return CloudPlusService;
}());
export { CloudPlusService };
