// @farris/mobile-command-services/esm5/lib/js-bridge-service/js-bridge-stratege.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/js-bridge-service/js-bridge-stratege.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsBridgeStrategyFactory } from './js-bridge-strategy-factory';
import { Subject } from 'rxjs';
var JsBridgeService = /** @class */ (function () {
    function JsBridgeService() {
        this.jsBridgeStrategyFactory = new JsBridgeStrategyFactory();
        this.init();
    }
    /**
     * @return {?}
     */
    JsBridgeService.prototype.getSysInfo = /**
     * @return {?}
     */
    function () {
        return this.getJsBridgeService(this.strategy.getSysInfo);
    };
    /**
     * @private
     * @return {?}
     */
    JsBridgeService.prototype.init = /**
     * @private
     * @return {?}
     */
    function () {
        this.strategy = this.jsBridgeStrategyFactory.create();
        window['JsBridgeService'] = this;
    };
    /**
     * @param {?} options
     * @return {?}
     */
    JsBridgeService.prototype.takePhoto = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return this.getJsBridgeService(this.strategy.takePhoto, options);
    };
    /**
     * @return {?}
     */
    JsBridgeService.prototype.scanQRCode = /**
     * @return {?}
     */
    function () {
        return this.getJsBridgeService(this.strategy.scanQRCode);
    };
    /**
     * @return {?}
     */
    JsBridgeService.prototype.getLocation = /**
     * @return {?}
     */
    function () {
        return this.getJsBridgeService(this.strategy.getLocation);
    };
    /**
     * @param {?} number
     * @param {?} content
     * @return {?}
     */
    JsBridgeService.prototype.sendSMS = /**
     * @param {?} number
     * @param {?} content
     * @return {?}
     */
    function (number, content) {
        return this.getJsBridgeService(this.strategy.sendSMS, number, content);
    };
    /**
     * @param {?} number
     * @param {?} content
     * @return {?}
     */
    JsBridgeService.prototype.openSendSMS = /**
     * @param {?} number
     * @param {?} content
     * @return {?}
     */
    function (number, content) {
        return this.getJsBridgeService(this.strategy.openSendSMS, number, content);
    };
    /**
     * @param {?} number
     * @return {?}
     */
    JsBridgeService.prototype.tel = /**
     * @param {?} number
     * @return {?}
     */
    function (number) {
        return this.getJsBridgeService(this.strategy.tel, number);
    };
    /**
     * @param {?} options
     * @return {?}
     */
    JsBridgeService.prototype.selectPictureFromAlbum = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return this.getJsBridgeService(this.strategy.selectPictureFromAlbum, options);
    };
    /**
     * @param {?} options
     * @return {?}
     */
    JsBridgeService.prototype.takeVideo = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return this.getJsBridgeService(this.strategy.takeVideo, options);
    };
    /**
     * @return {?}
     */
    JsBridgeService.prototype.closeWindow = /**
     * @return {?}
     */
    function () {
        return this.getJsBridgeService(this.strategy.closeWindow);
    };
    /**
     * @param {?} title
     * @return {?}
     */
    JsBridgeService.prototype.setTitles = /**
     * @param {?} title
     * @return {?}
     */
    function (title) {
        return this.getJsBridgeService(this.strategy.setTitles, title);
    };
    /**
     * @param {?=} options
     * @return {?}
     */
    JsBridgeService.prototype.selectFile = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        return this.getJsBridgeService(this.strategy.selectFile, options);
    };
    /**
     * @param {?} options
     * @return {?}
     */
    JsBridgeService.prototype.downloadFile = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return this.getJsBridgeService(this.strategy.downloadFile, options);
    };
    /**
     * @param {?=} fn
     * @return {?}
     */
    JsBridgeService.prototype.setOriginGoback = /**
     * @param {?=} fn
     * @return {?}
     */
    function (fn) {
        return this.getJsBridgeService(this.strategy.setOriginGoback, fn);
    };
    /**
     * @param {?=} fn
     * @return {?}
     */
    JsBridgeService.prototype.pushOriginGoback = /**
     * @param {?=} fn
     * @return {?}
     */
    function (fn) {
        return this.getJsBridgeService(this.strategy.pushOriginGoback, fn);
    };
    /**
     * @param {?=} fn
     * @return {?}
     */
    JsBridgeService.prototype.popOriginGoback = /**
     * @param {?=} fn
     * @return {?}
     */
    function (fn) {
        return this.getJsBridgeService(this.strategy.popOriginGoback, fn);
    };
    /**
     * @param {?=} fn
     * @return {?}
     */
    JsBridgeService.prototype.playShortVideo = /**
     * @param {?=} fn
     * @return {?}
     */
    function (fn) {
        return this.getJsBridgeService(this.strategy.playShortVideo, fn);
    };
    /**
     * @private
     * @param {?} method
     * @param {...?} values
     * @return {?}
     */
    JsBridgeService.prototype.getJsBridgeService = /**
     * @private
     * @param {?} method
     * @param {...?} values
     * @return {?}
     */
    function (method) {
        var values = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            values[_i - 1] = arguments[_i];
        }
        /** @type {?} */
        var subject = new Subject();
        // 已经存在方法实例即结束返回
        if (window.top['FARRIS_JSBridge'] === '已加载') {
            return method.apply(void 0, tslib_1.__spread(values));
        }
        // 判断是否为云上协同环境
        if (window.top['FARRIS_JSBridge'] === '云上协同') {
            if (window.top['iCityBridge']) {
                return method.apply(void 0, tslib_1.__spread(values));
            }
            else {
                /** @type {?} */
                var ccwork = window.top.document.createElement("script");
                ccwork.src = 'https://ccwork.oss.cn-north-3.inspurcloudoss.com/h5/ccworkJsbridge.js';
                ccwork.onload = (/**
                 * @return {?}
                 */
                function () {
                    window.top['iCityBridge'].init((/**
                     * @return {?}
                     */
                    function () {
                        method.apply(void 0, tslib_1.__spread(values)).subscribe((/**
                         * @param {?} e
                         * @return {?}
                         */
                        function (e) { subject.next(e); }));
                    }));
                });
                window.top.document.head.appendChild(ccwork);
            }
        }
        // 当前没有针对云+的判断，前置环境不满足即默认云+环境
        if (window.top['FARRIS_JSBridge'] === '云加') {
            System
                .import('/platform/common/web/imp.js')
                .then((/**
             * @param {?} e
             * @return {?}
             */
            function (e) {
                window.top['FARRIS_JSBridge'] = '已加载';
                method.apply(void 0, tslib_1.__spread(values)).subscribe((/**
                 * @param {?} e
                 * @return {?}
                 */
                function (e) { subject.next(e); }));
            }));
        }
        return subject;
    };
    return JsBridgeService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    JsBridgeService.prototype.strategy;
    /**
     * @type {?}
     * @private
     */
    JsBridgeService.prototype.jsBridgeStrategyFactory;
}
export { JsBridgeService };
