// @farris/mobile-vue-adapter/esm5/lib/component-extenders/filter-extender.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component-extenders/filter-extender.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 列表组件扩展
 */
var /**
 * 列表组件扩展
 */
FilterExtender = /** @class */ (function () {
    function FilterExtender(filterInstance, viewModelContext) {
        this.filterInstance = filterInstance;
        this.viewModelContext = viewModelContext;
    }
    /**
     * 扩展组件
     */
    /**
     * 扩展组件
     * @return {?}
     */
    FilterExtender.prototype.extend = /**
     * 扩展组件
     * @return {?}
     */
    function () {
        /** @type {?} */
        var self = this;
        // 清空uistate的筛选条件
        this.filterInstance.$on('clear', (/**
         * @return {?}
         */
        function () {
            self.viewModelContext.uiState.setPropertyValue('filterState', { filters: null });
        }));
        // 给uistate的赋值筛选条件
        // this.filterInstance.$on('change', (data) => {
        //   //self.viewModelContext.uiState.setPropertyValue('filterState', { filters: data.FilterConditions });
        // });
        // 给uistate的赋值筛选条件
        this.filterInstance.$on('search', (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            self.viewModelContext.repository.setPaginationInfo({
                pageIndex: 1
            });
            self.viewModelContext.uiState.setPropertyValue('filterState', { filters: data.FilterConditions });
            self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
        }));
        // 监听确定时间出发查询
        this.filterInstance.$on('confirm', (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            self.viewModelContext.repository.setPaginationInfo({
                pageIndex: 1
            });
            self.viewModelContext.uiState.setPropertyValue('filterState', { filters: data.FilterConditions });
            self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
        }));
    };
    /**
     * 清理状态
     */
    /**
     * 清理状态
     * @return {?}
     */
    FilterExtender.prototype.dispose = /**
     * 清理状态
     * @return {?}
     */
    function () {
        this.filterInstance.$off('clear');
        this.filterInstance.$off('change');
        this.filterInstance.$off('confirm');
    };
    return FilterExtender;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    FilterExtender.prototype.filterInstance;
    /**
     * @type {?}
     * @private
     */
    FilterExtender.prototype.viewModelContext;
}
export { FilterExtender };
