// @farris/mobile-vue-adapter/esm5/lib/component-extenders/listview-extender.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component-extenders/listview-extender.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/**
 * 列表组件扩展
 */
var /**
 * 列表组件扩展
 */
ListviewExtender = /** @class */ (function () {
    function ListviewExtender(listviewInstance, viewModelContext) {
        this.listviewInstance = listviewInstance;
        this.viewModelContext = viewModelContext;
    }
    /**
     * 扩展组件
     */
    /**
     * 扩展组件
     * @return {?}
     */
    ListviewExtender.prototype.extend = /**
     * 扩展组件
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var self = this;
        this.viewModelContext.appContext.eventBus.subscribe(this.viewModelContext.id + "-listview-extend", (/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value.type === 'listviewFinished') {
                _this.viewModelContext.uiState.setPropertyValue('listaviewState', tslib_1.__assign({}, _this.viewModelContext.uiState['listaviewState'], { loadMoreFinished: true }));
            }
            if (value.type === 'listviewUnFinished') {
                _this.viewModelContext.uiState.setPropertyValue('listaviewState', tslib_1.__assign({}, _this.viewModelContext.uiState['listaviewState'], { loadMoreFinished: false }));
            }
            if (value.type === 'loadDataFinished') {
                _this.viewModelContext.uiState.setPropertyValue('listaviewState', tslib_1.__assign({}, _this.viewModelContext.uiState['listaviewState'], { refreshLoading: false, loadMoreLoading: false }));
            }
        }));
        // 当前行
        this.listviewInstance.$on('selectChange', (/**
         * @param {?} selectedData
         * @return {?}
         */
        function (selectedData) {
            /** @type {?} */
            var selectedId = selectedData.data.id;
            self.viewModelContext.bindingData.getList().setCurrentId(selectedId, true, true);
        }));
        // 批量选择
        this.listviewInstance.$on('itemClick', (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var selectedItems = self.listviewInstance.getSelectDataList();
            /** @type {?} */
            var selectedIds = [];
            if (selectedItems && selectedItems.length > 0) {
                selectedIds = selectedItems.map((/**
                 * @param {?} selectedItem
                 * @return {?}
                 */
                function (selectedItem) {
                    return selectedItem.id;
                }));
            }
            self.viewModelContext.uiState.setPropertyValue('selectedIds', selectedIds);
        }));
        // 监听下拉刷新组件
        this.listviewInstance.$on('refresh', (/**
         * @return {?}
         */
        function () {
            self.viewModelContext.repository.setPaginationInfo({
                pageIndex: 1
            });
            self.viewModelContext.uiState.setPropertyValue('listaviewState', tslib_1.__assign({}, self.viewModelContext.uiState['listaviewState'], { refreshLoading: true, loadMoreLoading: true }));
            self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
        }));
        // 监听是否走查询
        this.listviewInstance.$on('load', (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var loading = self.viewModelContext.uiState['listaviewState']['loadMoreLoading'];
            if (!loading) {
                self.viewModelContext.uiState.setPropertyValue('listaviewState', tslib_1.__assign({}, self.viewModelContext.uiState['listaviewState'], { loadMoreLoading: true }));
                /** @type {?} */
                var pageIndex = self.viewModelContext.repository.paginationInfo && self.viewModelContext.repository.paginationInfo.pageIndex;
                if (!pageIndex) {
                    pageIndex = 0;
                }
                self.viewModelContext.repository.setPaginationInfo({
                    pageIndex: pageIndex += 1
                });
                self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
            }
        }));
    };
    /**
     * 清理状态
     */
    /**
     * 清理状态
     * @return {?}
     */
    ListviewExtender.prototype.dispose = /**
     * 清理状态
     * @return {?}
     */
    function () {
        this.listviewInstance.$off('selectChange');
        this.listviewInstance.$off('itemClick');
        this.listviewInstance.$off('load');
        this.viewModelContext.appContext.eventBus.unsubscribe(this.viewModelContext.id + "-listview-extend");
    };
    return ListviewExtender;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    ListviewExtender.prototype.listviewInstance;
    /**
     * @type {?}
     * @private
     */
    ListviewExtender.prototype.viewModelContext;
}
export { ListviewExtender };
