// @farris/mobile-vue-adapter/esm5/lib/component-extenders/map-extender.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component-extenders/map-extender.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 列表组件扩展
 */
var /**
 * 列表组件扩展
 */
MapExtender = /** @class */ (function () {
    function MapExtender(mapInstance, viewModelContext) {
        this.mapInstance = mapInstance;
        this.viewModelContext = viewModelContext;
    }
    /**
     * 扩展组件
     */
    /**
     * 扩展组件
     * @return {?}
     */
    MapExtender.prototype.extend = /**
     * 扩展组件
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var self = this;
        // 当前行
        this.mapInstance.$on('markMap', (/**
         * @param {?} selectedData
         * @return {?}
         */
        function (selectedData) {
            // 获取经纬度
            /** @type {?} */
            var lnglat = selectedData.lnglat;
            /** @type {?} */
            var lng = lnglat.lng;
            /** @type {?} */
            var lat = lnglat.lat;
            // 获取经纬度路径
            /** @type {?} */
            var lngPath = _this.mapInstance.lngPath;
            /** @type {?} */
            var latPath = _this.mapInstance.latPath;
            // this.mapInstance.center = [lng, lat];
            // 赋值经纬度
            _this.viewModelContext.bindingData.setValue(_this.getBindingPathArray().concat(lngPath.split('/')), lng, true, true);
            _this.viewModelContext.bindingData.setValue(_this.getBindingPathArray().concat(latPath.split('/')), lat, true, true);
        }));
    };
    /**
     * 清理状态
     */
    /**
     * 清理状态
     * @return {?}
     */
    MapExtender.prototype.dispose = /**
     * 清理状态
     * @return {?}
     */
    function () {
        this.mapInstance.$off('markMap');
    };
    /**
     * @private
     * @return {?}
     */
    MapExtender.prototype.getBindingPathArray = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var path = this.viewModelContext.bindingData.bindingPath;
        if (path) {
            return path.split('/').filter((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return n !== ''; }));
        }
        return [];
    };
    return MapExtender;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    MapExtender.prototype.mapInstance;
    /**
     * @type {?}
     * @private
     */
    MapExtender.prototype.viewModelContext;
}
export { MapExtender };
