// @farris/mobile-vue-adapter/esm5/lib/component-extenders/reflesh-extender.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component-extenders/reflesh-extender.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/**
 * 列表组件扩展
 */
var /**
 * 列表组件扩展
 */
RefleshExtender = /** @class */ (function () {
    function RefleshExtender(refleshInstance, viewModelContext) {
        this.refleshInstance = refleshInstance;
        this.viewModelContext = viewModelContext;
    }
    /**
     * 扩展组件
     */
    /**
     * 扩展组件
     * @return {?}
     */
    RefleshExtender.prototype.extend = /**
     * 扩展组件
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var self = this;
        // 监听下拉刷新组件
        this.refleshInstance.$on('input', (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                self.viewModelContext.repository.setPaginationInfo({
                    pageIndex: 1
                });
                self.viewModelContext.uiState.setPropertyValue('listaviewState', tslib_1.__assign({}, self.viewModelContext.uiState['listaviewState'], { loadMoreLoading: true }));
                self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
            }
            self.viewModelContext.appContext.eventBus.subscribe(_this.viewModelContext.id + "-reflesh-extend", (/**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                if (value.type === 'loadDataFinished') {
                    self.viewModelContext.uiState.setPropertyValue('listaviewState', tslib_1.__assign({}, self.viewModelContext.uiState['listaviewState'], { refreshLoading: false, loadMoreLoading: false }));
                }
            }));
        }));
    };
    /**
     * 清理状态
     */
    /**
     * 清理状态
     * @return {?}
     */
    RefleshExtender.prototype.dispose = /**
     * 清理状态
     * @return {?}
     */
    function () {
        this.refleshInstance.$off('input');
        this.viewModelContext.appContext.eventBus.unsubscribe(this.viewModelContext.id + "-reflesh-extend");
    };
    return RefleshExtender;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    RefleshExtender.prototype.refleshInstance;
    /**
     * @type {?}
     * @private
     */
    RefleshExtender.prototype.viewModelContext;
}
export { RefleshExtender };
