// @farris/mobile-vue-adapter/esm5/lib/component-extenders/search-extender.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component-extenders/search-extender.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 列表组件扩展
 */
var /**
 * 列表组件扩展
 */
SearchExtender = /** @class */ (function () {
    function SearchExtender(searchInstance, viewModelContext) {
        this.searchInstance = searchInstance;
        this.viewModelContext = viewModelContext;
    }
    /**
     * 扩展组件
     */
    /**
     * 扩展组件
     * @return {?}
     */
    SearchExtender.prototype.extend = /**
     * 扩展组件
     * @return {?}
     */
    function () {
        /** @type {?} */
        var self = this;
        // 监听搜索组件
        this.searchInstance.$on('cancel', (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            //  清空走查询
            self.viewModelContext.repository.setPaginationInfo({
                pageIndex: 1
            });
            self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
        }));
        this.searchInstance.$on('search', (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            // 如果是确定类型，即走查询
            self.viewModelContext.repository.setPaginationInfo({
                pageIndex: 1
            });
            self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
        }));
    };
    /**
     * 清理状态
     */
    /**
     * 清理状态
     * @return {?}
     */
    SearchExtender.prototype.dispose = /**
     * 清理状态
     * @return {?}
     */
    function () {
        this.searchInstance.$off('cancel');
        this.searchInstance.$off('search');
    };
    return SearchExtender;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    SearchExtender.prototype.searchInstance;
    /**
     * @type {?}
     * @private
     */
    SearchExtender.prototype.viewModelContext;
}
export { SearchExtender };
