// @edp-aif/common-api/esm2015/lib/entity/impl/AifConfigEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
import { Entity } from './Entity';
export class AifConfigEntity extends Entity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.content = null;
        this.tag = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tag", this.tag);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
        this.tag = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tag");
    }
}
if (false) {
    /** @type {?} */
    AifConfigEntity.prototype.content;
    /** @type {?} */
    AifConfigEntity.prototype.tag;
}
