// @edp-aif/common-api/esm2015/lib/entity/impl/BizFlowchartRefMetadataEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BizFlowchartRefEntity } from './BizFlowchartRefEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
export class BizFlowchartRefMetadataEntity extends BizFlowchartRefEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.refAppCode = null;
        this.refSuCode = null;
        this.refBoId = null;
        this.bizFlowchartAppCode = null;
        this.bizFlowchartSuCode = null;
        this.bizFlowchartBoId = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "refAppCode", this.refAppCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "refSuCode", this.refSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "refBoId", this.refBoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartAppCode", this.bizFlowchartAppCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartSuCode", this.bizFlowchartSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartBoId", this.bizFlowchartBoId);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.refAppCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "refAppCode");
        this.refSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "refSuCode");
        this.refBoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "refBoId");
        this.bizFlowchartAppCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartAppCode");
        this.bizFlowchartSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartSuCode");
        this.bizFlowchartBoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartBoId");
    }
}
if (false) {
    /** @type {?} */
    BizFlowchartRefMetadataEntity.prototype.refAppCode;
    /** @type {?} */
    BizFlowchartRefMetadataEntity.prototype.refSuCode;
    /** @type {?} */
    BizFlowchartRefMetadataEntity.prototype.refBoId;
    /** @type {?} */
    BizFlowchartRefMetadataEntity.prototype.bizFlowchartAppCode;
    /** @type {?} */
    BizFlowchartRefMetadataEntity.prototype.bizFlowchartSuCode;
    /** @type {?} */
    BizFlowchartRefMetadataEntity.prototype.bizFlowchartBoId;
}
