// @edp-aif/common-api/esm2015/lib/entity/impl/BizProcessEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { P2PMetadataEntity } from './P2PMetadataEntity';
import { CallType } from '../CallType';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * 业务流程实体类。
 * @abstract
 */
export class BizProcessEntity extends P2PMetadataEntity {
    constructor() {
        super(...arguments);
        this.callType = CallType.SYNC;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.bizEventId = null;
        this.bizActionId = null;
        this.isEnabled = false;
        this.callType = CallType.SYNC;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventId", this.bizEventId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionId", this.bizActionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "callType", this.callType);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.bizEventId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventId");
        this.bizActionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionId");
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.callType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "callType");
    }
}
if (false) {
    /** @type {?} */
    BizProcessEntity.prototype.bizEventId;
    /** @type {?} */
    BizProcessEntity.prototype.bizActionId;
    /** @type {?} */
    BizProcessEntity.prototype.isEnabled;
    /** @type {?} */
    BizProcessEntity.prototype.callType;
}
