// @edp-aif/common-api/esm2015/lib/entity/impl/CreationProcessEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BizProcessEntity } from './BizProcessEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * 生单流程实体类。
 */
export class CreationProcessEntity extends BizProcessEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.creationRuleId = null;
        this.reverseWriteRuleId = null;
        this.reverseWriteRuleClassName = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleId", this.reverseWriteRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleClassName", this.reverseWriteRuleClassName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.reverseWriteRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleId");
        this.reverseWriteRuleClassName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleClassName");
    }
}
if (false) {
    /** @type {?} */
    CreationProcessEntity.prototype.creationRuleId;
    /** @type {?} */
    CreationProcessEntity.prototype.reverseWriteRuleId;
    /** @type {?} */
    CreationProcessEntity.prototype.reverseWriteRuleClassName;
}
