// @edp-aif/common-api/esm2015/lib/entity/impl/CreationRuleEntityEx.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CreationRuleEntity } from './CreationRuleEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
export class CreationRuleEntityEx extends CreationRuleEntity {
    /**
     * @param {?=} entity
     */
    constructor(entity) {
        super();
        if (entity != null) {
            this.id = entity.id;
            this.appCode = entity.appCode;
            this.boId = entity.boId;
            this.createdTime = entity.createdTime;
            this.creator = entity.creator;
            this.description = entity.description;
            this.isEnabled = entity.isEnabled;
            this.lastModifiedTime = entity.lastModifiedTime;
            this.lastModifier = entity.lastModifier;
            this.mappingComponentId = entity.mappingComponentId;
            this.mappingComponentMetadataId = entity.mappingComponentMetadataId;
            this.mappingRuleId = entity.mappingRuleId;
            this.mappingRuleMetadataId = entity.mappingRuleMetadataId;
            this.name = entity.name;
            this.persistenceVersion = entity.persistenceVersion;
            this.sourceAppCode = entity.sourceAppCode;
            this.sourceBeId = entity.sourceBeId;
            this.sourceBeMetadataId = entity.sourceBeMetadataId;
            //this.sourceBeType = entity.sourceBeType;
            this.sourceBoId = entity.sourceBoId;
            this.sourceSuCode = entity.sourceSuCode;
            this.sourceVoId = entity.sourceVoId;
            this.sourceVoMetadataId = entity.sourceVoMetadataId;
            //this.sourceVoType = entity.sourceVoType;
            this.suCode = entity.suCode;
            this.targetAppCode = entity.targetAppCode;
            this.targetBeId = entity.targetBeId;
            this.targetBeMetadataId = entity.targetBeMetadataId;
            //this.targetBeType = entity.targetBeType;
            this.targetBoId = entity.targetBoId;
            this.targetSuCode = entity.targetSuCode;
            this.targetVoId = entity.targetVoId;
            this.targetVoMetadataId = entity.targetVoMetadataId;
            this.targetVoProxyId = entity.targetVoProxyId;
            this.targetVoProxyMetadataId = entity.targetVoProxyMetadataId;
            //this.targetVoProxyType = entity.targetVoProxyType;
            //this.targetVoType = entity.targetVoType;
        }
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.sourceBeName = null;
        this.targetBeName = null;
        this.sourceVoName = null;
        this.targetVoName = null;
        this.targetVoProxyName = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBeName", this.sourceBeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetBeName", this.targetBeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoName", this.sourceVoName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoName", this.targetVoName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyName", this.targetVoProxyName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.sourceBeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBeName");
        this.targetBeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBeName");
        this.sourceVoName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoName");
        this.targetVoName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoName");
        this.targetVoProxyName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyName");
    }
}
if (false) {
    /** @type {?} */
    CreationRuleEntityEx.prototype.sourceBeName;
    /** @type {?} */
    CreationRuleEntityEx.prototype.targetBeName;
    /** @type {?} */
    CreationRuleEntityEx.prototype.sourceVoName;
    /** @type {?} */
    CreationRuleEntityEx.prototype.targetVoName;
    /** @type {?} */
    CreationRuleEntityEx.prototype.targetVoProxyName;
}
