// @edp-aif/common-api/esm2015/lib/entity/impl/Entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl } from '../../JsonConvertableImpl';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * 实体类接口。
 * @abstract
 */
export class Entity extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.id = null;
        this.creator = null;
        this.createdTime = null;
        this.lastModifier = null;
        this.lastModifiedTime = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "id", this.id);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creator", this.creator);
        BpmnModelJsonHelper.SetAttributeValue(obj, "createdTime", this.createdTime);
        BpmnModelJsonHelper.SetAttributeValue(obj, "lastModifier", this.lastModifier);
        BpmnModelJsonHelper.SetAttributeValue(obj, "lastModifiedTime", this.lastModifiedTime);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.id = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "id");
        this.creator = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creator");
        this.createdTime = BpmnModelJsonHelper.GetAttributeValueAsDateTime(obj, "createdTime");
        this.lastModifier = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "lastModifier");
        this.lastModifiedTime = BpmnModelJsonHelper.GetAttributeValueAsDateTime(obj, "lastModifiedTime");
    }
}
if (false) {
    /**
     * 唯一标识。
     * @type {?}
     */
    Entity.prototype.id;
    /** @type {?} */
    Entity.prototype.creator;
    /** @type {?} */
    Entity.prototype.createdTime;
    /** @type {?} */
    Entity.prototype.lastModifier;
    /** @type {?} */
    Entity.prototype.lastModifiedTime;
}
