// @edp-aif/common-api/esm2015/lib/entity/impl/GeneralTextEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Entity } from './Entity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
export class GeneralTextEntity extends Entity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.text = null;
        this.tag = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "text", this.text);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tag", this.tag);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.text = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "text");
        this.tag = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tag");
    }
}
if (false) {
    /** @type {?} */
    GeneralTextEntity.prototype.text;
    /** @type {?} */
    GeneralTextEntity.prototype.tag;
}
