// @edp-aif/common-api/esm2015/lib/entity/impl/MetadataEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Entity } from './Entity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * 元数据实体类接口。
 * @abstract
 */
export class MetadataEntity extends Entity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.appCode = null;
        this.suCode = null;
        this.boId = null;
        this.persistenceVersion = null;
        this.name = null;
        this.description = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "appCode", this.appCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "boId", this.boId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "persistenceVersion", this.persistenceVersion);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "description", this.description);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.appCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "appCode");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.boId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "boId");
        this.persistenceVersion = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "persistenceVersion");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.description = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "description");
    }
}
if (false) {
    /** @type {?} */
    MetadataEntity.prototype.appCode;
    /** @type {?} */
    MetadataEntity.prototype.suCode;
    /** @type {?} */
    MetadataEntity.prototype.boId;
    /** @type {?} */
    MetadataEntity.prototype.persistenceVersion;
    /** @type {?} */
    MetadataEntity.prototype.name;
    /** @type {?} */
    MetadataEntity.prototype.description;
}
