// @edp-aif/common-api/esm2015/lib/entity/impl/Participant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from "../../BpmnModelJsonHelper";
import { JsonConvertableImpl } from "../../JsonConvertableImpl";
import { ParticipantType } from "./ParticipantType";
import { PositionParticipant } from "./PositionParticipant";
import { UserParticipant } from "./UserParticipant";
export class Participant extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.participantType = ParticipantType.UserParticipant;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.participantType = ParticipantType.UserParticipant;
        this.userParticipant = null;
        this.positionParticipant = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "participantType", this.participantType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "userParticipant", this.userParticipant);
        BpmnModelJsonHelper.SetAttributeValue(obj, "positionParticipant", this.positionParticipant);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.participantType = ParticipantType[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "participantType")];
        this.userParticipant = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "userParticipant", UserParticipant);
        this.positionParticipant = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "positionParticipant", PositionParticipant);
    }
}
if (false) {
    /** @type {?} */
    Participant.prototype.participantType;
    /** @type {?} */
    Participant.prototype.userParticipant;
    /** @type {?} */
    Participant.prototype.positionParticipant;
}
