// @edp-aif/common-api/esm2015/lib/entity/impl/PositionParticipant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from "../../BpmnModelJsonHelper";
import { JsonConvertableImpl } from "../../JsonConvertableImpl";
export class PositionParticipant extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.orgType = "Sys";
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.positionId = null;
        this.positionName = null;
        this.positionType = null;
        this.orgId = null;
        this.orgName = null;
        this.orgType = null;
        this.orgIdExpression = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "positionId", this.positionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "positionName", this.positionName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "positionType", this.positionType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgId", this.orgId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgName", this.orgName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgType", this.orgType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgIdExpression", this.orgIdExpression);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.positionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "positionId");
        this.positionName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "positionName");
        this.positionType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "positionType");
        this.orgId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgId");
        this.orgName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgName");
        this.orgType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgType");
        this.orgIdExpression = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgIdExpression");
    }
}
if (false) {
    /** @type {?} */
    PositionParticipant.prototype.positionId;
    /** @type {?} */
    PositionParticipant.prototype.positionName;
    /** @type {?} */
    PositionParticipant.prototype.positionType;
    /** @type {?} */
    PositionParticipant.prototype.orgId;
    /** @type {?} */
    PositionParticipant.prototype.orgName;
    /** @type {?} */
    PositionParticipant.prototype.orgType;
    /** @type {?} */
    PositionParticipant.prototype.orgIdExpression;
}
