// @edp-aif/common-api/esm2015/lib/entity/impl/TaskHandlingMenuArgs.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from "../../BpmnModelJsonHelper";
import { JsonConvertableImpl } from "../../JsonConvertableImpl";
import { TaskHandlingMenuParameter } from "./TaskHandlingMenuParameter";
export class TaskHandlingMenuArgs extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.parameters = [];
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.menuId = null;
        this.menuName = null;
        this.parameters = [];
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "menuId", this.menuId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "menuName", this.menuName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "parameters", this.parameters);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.menuId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "menuId");
        this.menuName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "menuName");
        this.parameters = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "parameters", TaskHandlingMenuParameter);
    }
}
if (false) {
    /** @type {?} */
    TaskHandlingMenuArgs.prototype.menuId;
    /** @type {?} */
    TaskHandlingMenuArgs.prototype.menuName;
    /** @type {?} */
    TaskHandlingMenuArgs.prototype.parameters;
}
