// @edp-aif/common-api/esm2015/lib/entity/impl/TaskHandlingMenuParameter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from "../../BpmnModelJsonHelper";
import { JsonConvertableImpl } from "../../JsonConvertableImpl";
import { TaskHandlingMenuParameterMode } from "./TaskHandlingMenuParameterMode";
export class TaskHandlingMenuParameter extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.parameterMode = TaskHandlingMenuParameterMode.CONSTANT;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.parameterMode = TaskHandlingMenuParameterMode.CONSTANT;
        this.code = null;
        this.constantValue = null;
        this.fieldContent = null;
        this.expressionContent = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "parameterMode", this.parameterMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "constantValue", this.constantValue);
        BpmnModelJsonHelper.SetAttributeValue(obj, "fieldContent", this.fieldContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "expressionContent", this.expressionContent);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.parameterMode = TaskHandlingMenuParameterMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "parameterMode")];
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.constantValue = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "constantValue");
        this.fieldContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "fieldContent");
        this.expressionContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "expressionContent");
    }
}
if (false) {
    /** @type {?} */
    TaskHandlingMenuParameter.prototype.parameterMode;
    /** @type {?} */
    TaskHandlingMenuParameter.prototype.code;
    /** @type {?} */
    TaskHandlingMenuParameter.prototype.constantValue;
    /** @type {?} */
    TaskHandlingMenuParameter.prototype.fieldContent;
    /** @type {?} */
    TaskHandlingMenuParameter.prototype.expressionContent;
}
