// @edp-aif/common-api/esm2015/lib/entity/impl/TaskCenterContent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from "../../BpmnModelJsonHelper";
import { JsonConvertableImpl } from "../../JsonConvertableImpl";
import { Participant } from "./Participant";
import { TaskHandlingMenuArgs } from "./TaskHandlingMenuArgs";
export class TaskCenterContent extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.participants = [];
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.participants = [];
        this.taskHandlingMenuArgs = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "participants", this.participants);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskHandlingMenuArgs", this.taskHandlingMenuArgs);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.participants = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "participants", Participant);
        this.taskHandlingMenuArgs = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "taskHandlingMenuArgs", TaskHandlingMenuArgs);
    }
}
if (false) {
    /** @type {?} */
    TaskCenterContent.prototype.participants;
    /** @type {?} */
    TaskCenterContent.prototype.taskHandlingMenuArgs;
}
