// @edp-aif/common-api/esm2015/lib/entity/impl/UserParticipant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from "../../BpmnModelJsonHelper";
import { JsonConvertableImpl } from "../../JsonConvertableImpl";
export class UserParticipant extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.orgType = "Sys";
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.userId = null;
        this.userName = null;
        this.orgId = null;
        this.orgName = null;
        this.orgType = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "userId", this.userId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "userName", this.userName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgId", this.orgId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgName", this.orgName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgType", this.orgType);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.userId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "userId");
        this.userName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "userName");
        this.orgId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgId");
        this.orgName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgName");
        this.orgType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgType");
    }
}
if (false) {
    /** @type {?} */
    UserParticipant.prototype.userId;
    /** @type {?} */
    UserParticipant.prototype.userName;
    /** @type {?} */
    UserParticipant.prototype.orgId;
    /** @type {?} */
    UserParticipant.prototype.orgName;
    /** @type {?} */
    UserParticipant.prototype.orgType;
}
