// @farris/devkit/esm2015/lib/binding-data/binding_list_type_factory.js
import { BindingList } from "./binding_list";
/**
 * BindingListTypeFactory
 */
export class BindingListTypeFactory {
    static create(bindingProperties) {
        const bindingListType = this.getType(bindingProperties);
        const bindingList = new bindingListType(bindingProperties);
        return bindingList;
    }
    /**
     * 创建BindingList模板类
     * @param bindingProperties
     * @returns
     */
    static createType(bindingProperties) {
        const bindingListType = class BindingListType extends BindingList {
            constructor(properties) {
                super(properties);
            }
        };
        const bindingListTypePrototype = bindingListType.prototype;
        this.extendProperties(bindingListTypePrototype, bindingProperties);
        return bindingListType;
    }
    /**
     * 扩展BindingList模板类属性
     * @param bindingListPrototype bindingList原型
     * @param bindingProperties 属性数组
     */
    static extendProperties(bindingListPrototype, bindingProperties) {
        bindingProperties.forEach((bindingProperty) => {
            const propertyName = bindingProperty.name;
            Object.defineProperty(bindingListPrototype, propertyName, {
                get: function () {
                    return this.currentItem[propertyName];
                }
            });
        });
    }
    /**
     * 获取缓存的bindingList模板类
     * @param bindingProperties bindingList属性
     * @returns
     */
    static getType(bindingProperties) {
        if (this.provider.has(bindingProperties)) {
            return this.provider.get(bindingProperties);
        }
        const bindingListType = this.createType(bindingProperties);
        this.provider.set(bindingProperties, bindingListType);
        return bindingListType;
    }
}
BindingListTypeFactory.provider = new Map();
