// @farris/ui-filter-panel/esm5/lib/components/singlenumbercomponent/singlenumber.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output, ViewChild } from '@angular/core';
import { NumberSpinnerComponent } from '@farris/ui-number-spinner';
var PanelSingleNumberComponent = /** @class */ (function () {
    function PanelSingleNumberComponent() {
        this.searchHandle = new EventEmitter();
        this.submitChange = new EventEmitter();
        this.canNull = true;
    }
    Object.defineProperty(PanelSingleNumberComponent.prototype, "item", {
        get: /**
         * @return {?}
         */
        function () {
            return this._item;
        },
        set: /**
         * @param {?} nr
         * @return {?}
         */
        function (nr) {
            this._item = nr;
            if (this._item) {
                if (!this._item.hasOwnProperty('value')) {
                    this._item.value = { value: null };
                }
                if (this._item.placeHolder) {
                    this.placeholder = this._item.placeHolder || '';
                }
            }
            this.precision = nr.control.precision;
            this.bigNumber = nr.control.bigNumber;
            if (!this.bigNumber) {
                this.min = -2147483648;
                this.max = 2147483647;
            }
            this.setSubmitChange();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PanelSingleNumberComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.numberText) {
            this.numberText.input.nativeElement.focus();
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PanelSingleNumberComponent.prototype.numberChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this._item.value.value = event;
        if (Object.keys(this._item.value).length === 0) {
            this._item.valueText = '';
        }
        else {
            this._item.valueText = '' + this._item.value.value;
        }
        this.setSubmitChange();
        this.searchHandle.emit();
    };
    /**
     * @return {?}
     */
    PanelSingleNumberComponent.prototype.setSubmitChange = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var val = this._item.value;
        if (Object.keys(val).length < 0 || val.value == null) {
            this.submitChange.emit(false);
        }
        else {
            this.submitChange.emit(true);
        }
    };
    PanelSingleNumberComponent.decorators = [
        { type: Component, args: [{
                    selector: 'panel-filter-single-number',
                    template: "<farris-number-spinner\r\n    class=\"f-utils-fill-flex-row\"\r\n    [(ngModel)]=\"item.value.value\"\r\n    [canNull]=\"canNull\"\r\n    [precision]=\"precision\"\r\n    [bigNumber]=\"bigNumber\"\r\n    [min]=\"min\" \r\n    [max]=\"max\"\r\n    [placeholder]=\"placeholder\"\r\n    (valueChange)=\"numberChange($event)\"\r\n    #numberText\r\n></farris-number-spinner>"
                }] }
    ];
    /** @nocollapse */
    PanelSingleNumberComponent.ctorParameters = function () { return []; };
    PanelSingleNumberComponent.propDecorators = {
        searchHandle: [{ type: Output }],
        submitChange: [{ type: Output }],
        item: [{ type: Input }],
        numberText: [{ type: ViewChild, args: ['numberText',] }]
    };
    return PanelSingleNumberComponent;
}());
export { PanelSingleNumberComponent };
if (false) {
    /** @type {?} */
    PanelSingleNumberComponent.prototype.searchHandle;
    /** @type {?} */
    PanelSingleNumberComponent.prototype.submitChange;
    /** @type {?} */
    PanelSingleNumberComponent.prototype.canNull;
    /** @type {?} */
    PanelSingleNumberComponent.prototype.number;
    /** @type {?} */
    PanelSingleNumberComponent.prototype.precision;
    /** @type {?} */
    PanelSingleNumberComponent.prototype.placeholder;
    /** @type {?} */
    PanelSingleNumberComponent.prototype.bigNumber;
    /** @type {?} */
    PanelSingleNumberComponent.prototype.max;
    /** @type {?} */
    PanelSingleNumberComponent.prototype.min;
    /** @type {?} */
    PanelSingleNumberComponent.prototype._item;
    /** @type {?} */
    PanelSingleNumberComponent.prototype.numberRange;
    /** @type {?} */
    PanelSingleNumberComponent.prototype.numberText;
}
