// @gsp-dip/data-imp-exp/esm5/lib/model/Model/DataExchangeElement.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var DataExchangeElement = /** @class */ (function () {
    function DataExchangeElement() {
        /**
         * 是否导出
         */
        this.Export = true;
        /**
         * 文件中字段名称
         */
        this.FileElementName = '';
        /**
         * (导出) 格式化数据的格式
         */
        this.Format = '';
        /**
         * (Excel) 列是否写保护
         */
        this.WriteProtect = false;
        /**
         * 序号
         */
        this.Order = 1;
        this.IsSum = false;
        /**
         * 计算表达式
         */
        this.Expression = '';
        /**
         * 所属数据交换对象Id
         */
        this.BelongDeObjectID = '';
        /**
         *  所属数据交换模型Id
         */
        this.BelongDeModelID = '';
    }
    /**
     * @param {?} ele
     * @return {?}
     */
    DataExchangeElement.copy = /**
     * @param {?} ele
     * @return {?}
     */
    function (ele) {
        /** @type {?} */
        var element = new DataExchangeElement();
        element.Export = ele.Export;
        element.ID = ele.ID;
        element.Code = ele.Code;
        element.Name = ele.Name;
        element.LabelID = ele.LabelID;
        element.MDataType = ele.MDataType;
        element.Source = ele.Source;
        element.IsKey = ele.IsKey;
        element.Format = ele.Format;
        element.FileElementName = ele.FileElementName;
        element.BelongDeObjectID = ele.BelongDeObjectID;
        element.BelongDeModelID = ele.BelongDeModelID;
        element.Expression = ele.Expression;
        element.WriteProtect = ele.WriteProtect;
        element.Order = ele.Order;
        element.Length = ele.Length;
        element.Precision = ele.Precision;
        return element;
    };
    return DataExchangeElement;
}());
export { DataExchangeElement };
if (false) {
    /**
     * 数据来源
     * @type {?}
     */
    DataExchangeElement.prototype.Source;
    /**
     * 是否导出
     * @type {?}
     */
    DataExchangeElement.prototype.Export;
    /**
     * 是否主键
     * @type {?}
     */
    DataExchangeElement.prototype.IsKey;
    /**
     * 文件中字段名称
     * @type {?}
     */
    DataExchangeElement.prototype.FileElementName;
    /**
     * (导出) 格式化数据的格式
     * @type {?}
     */
    DataExchangeElement.prototype.Format;
    /**
     * (Excel) 列是否写保护
     * @type {?}
     */
    DataExchangeElement.prototype.WriteProtect;
    /**
     * 序号
     * @type {?}
     */
    DataExchangeElement.prototype.Order;
    /** @type {?} */
    DataExchangeElement.prototype.IsSum;
    /**
     * 计算表达式
     * @type {?}
     */
    DataExchangeElement.prototype.Expression;
    /**
     * 所属数据交换对象Id
     * @type {?}
     */
    DataExchangeElement.prototype.BelongDeObjectID;
    /** @type {?} */
    DataExchangeElement.prototype.ID;
    /** @type {?} */
    DataExchangeElement.prototype.Code;
    /** @type {?} */
    DataExchangeElement.prototype.Name;
    /** @type {?} */
    DataExchangeElement.prototype.LabelID;
    /** @type {?} */
    DataExchangeElement.prototype.MDataType;
    /** @type {?} */
    DataExchangeElement.prototype.Length;
    /** @type {?} */
    DataExchangeElement.prototype.Precision;
    /**
     *  所属数据交换模型Id
     * @type {?}
     */
    DataExchangeElement.prototype.BelongDeModelID;
}
