// @gsp-dip/data-imp-exp/esm5/lib/model/Model/DataExchangeObject.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DataExchangeElement } from './DataExchangeElement';
import { DemSource } from './DemSource';
import { RepeatBasis } from '../InputStep/RepeatBasis/repeat-basis';
/**
 * 数据转换对象
 */
var /**
 * 数据转换对象
 */
DataExchangeObject = /** @class */ (function () {
    function DataExchangeObject() {
        /**
         * 父对象ID
         */
        this.ParentObjectId = '';
        /**
         * 文件中对象名称
         */
        this.FileObjectName = '';
        /**
         * 是否导出
         */
        this.Export = true;
        /**
         * 序号
         */
        this.Order = 1;
        /**
         * 对象字段集合
         */
        this.Elements = new Array();
        /**
         * 子对象集合
         */
        this.ChildrenDeo = new Array();
        /**
         * 数据来源
         */
        this.Source = new DemSource();
        /**
         * 重复依据
         */
        this.RepeatBases = new Array();
    }
    /**
     * copy
     */
    /**
     * copy
     * @param {?} obj
     * @return {?}
     */
    DataExchangeObject.copy = /**
     * copy
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var e_1, _a, e_2, _b, e_3, _c;
        /** @type {?} */
        var deo = new DataExchangeObject();
        deo.Export = obj.Export;
        deo.ID = obj.ID;
        deo.Code = obj.Code;
        deo.Name = obj.Name;
        deo.Source = obj.Source;
        deo.ParentObjectId = obj.ParentObjectId;
        deo.FileObjectName = obj.FileObjectName;
        deo.Order = obj.Order;
        deo.RefObjectName = obj.RefObjectName;
        deo.Elements = new Array();
        try {
            for (var _d = tslib_1.__values(obj.Elements), _e = _d.next(); !_e.done; _e = _d.next()) {
                var ele = _e.value;
                deo.Elements.push(DataExchangeElement.copy(ele));
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
            }
            finally { if (e_1) throw e_1.error; }
        }
        deo.ChildrenDeo = new Array();
        try {
            for (var _f = tslib_1.__values(obj.ChildrenDeo), _g = _f.next(); !_g.done; _g = _f.next()) {
                var item = _g.value;
                deo.ChildrenDeo.push(DataExchangeObject.copy(item));
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_g && !_g.done && (_b = _f.return)) _b.call(_f);
            }
            finally { if (e_2) throw e_2.error; }
        }
        deo.RepeatBases = new Array();
        try {
            for (var _h = tslib_1.__values(obj.RepeatBases), _j = _h.next(); !_j.done; _j = _h.next()) {
                var item = _j.value;
                deo.RepeatBases.push(RepeatBasis.copy(item));
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_j && !_j.done && (_c = _h.return)) _c.call(_h);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return deo;
    };
    return DataExchangeObject;
}());
/**
 * 数据转换对象
 */
export { DataExchangeObject };
if (false) {
    /**
     * 父对象ID
     * @type {?}
     */
    DataExchangeObject.prototype.ParentObjectId;
    /**
     * 文件中对象名称
     * @type {?}
     */
    DataExchangeObject.prototype.FileObjectName;
    /**
     * 是否导出
     * @type {?}
     */
    DataExchangeObject.prototype.Export;
    /**
     * 序号
     * @type {?}
     */
    DataExchangeObject.prototype.Order;
    /**
     * 对象字段集合
     * @type {?}
     */
    DataExchangeObject.prototype.Elements;
    /**
     * 子对象集合
     * @type {?}
     */
    DataExchangeObject.prototype.ChildrenDeo;
    /**
     * 数据来源
     * @type {?}
     */
    DataExchangeObject.prototype.Source;
    /**
     * 重复依据
     * @type {?}
     */
    DataExchangeObject.prototype.RepeatBases;
    /** @type {?} */
    DataExchangeObject.prototype.ID;
    /** @type {?} */
    DataExchangeObject.prototype.Code;
    /** @type {?} */
    DataExchangeObject.prototype.Name;
    /** @type {?} */
    DataExchangeObject.prototype.RefObjectName;
}
