// @gsp-wf/wf-process-editor/esm5/lib/views/execute-forms/variable-grid/variable-grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, EventEmitter, Inject, LOCALE_ID, Optional, Output, TemplateRef, ViewChild } from '@angular/core';
import { DatagridComponent } from '@farris/ui-datagrid';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../../i18n/pipe/wf-process-editor-locale.pipe';
var VariableGridComponent = /** @class */ (function () {
    function VariableGridComponent(cdr, localeId, resources, localePipe) {
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.cancel = new EventEmitter();
        this.submit = new EventEmitter();
        this.data = [];
        this.columns = [
            { field: 'key', width: 50, title: this.localePipe.transform('component.executeForms.key') },
            { field: 'name', width: 50, title: this.localePipe.transform('component.executeForms.name') }
        ];
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
    }
    /**
     * @return {?}
     */
    VariableGridComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    VariableGridComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var selectedRow = this.grid.selectedRow;
        console.log(selectedRow);
        if (!selectedRow) {
            this.submit.next(null);
        }
        else {
            this.submit.next(selectedRow.data);
        }
    };
    /**
     * @return {?}
     */
    VariableGridComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.cancel.next();
    };
    VariableGridComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-variable-grid',
                    template: "<div class=\"f-page\">\r\n  <div class=\"f-page-main f-utils-fill-flex-column h-100\">\r\n    <div class=\"f-utils-fill-flex-column farris-overflow-y-auto h-100\">\r\n      <farris-datagrid #grid [data]=\"data\" [columns]=\"columns\" [idField]=\"'key'\" [fit]=\"true\" [fitColumns]=\"true\"\r\n         [pagination]=\"false\" [showCheckbox]=\"false\" [multiSelect]=\"false\" [showLineNumber]=\"true\" lineNumberTitle=\"{{'component.executeForms.lineNumber' | wfLocale}}\"\r\n         [lineNumberWidth]=\"50\">\r\n      </farris-datagrid>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #buttonRef>\r\n    <button type=\"button\" class=\"btn btn-secondary btn-sm px-3\" (click)=\"clickCancel()\">{{'component.executeForms.clickCancel' | wfLocale}}</button>\r\n    <button type=\"button\" class=\"btn btn-primary btn-sm px-3 mx-2\" (click)=\"clickOK()\">{{'component.executeForms.clickOK' | wfLocale}}</button>\r\n</ng-template>\r\n",
                    providers: [
                        WfLocalePipe,
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    VariableGridComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
        { type: WfLocalePipe, decorators: [{ type: Optional }] }
    ]; };
    VariableGridComponent.propDecorators = {
        grid: [{ type: ViewChild, args: [DatagridComponent,] }],
        buttonRef: [{ type: ViewChild, args: ['buttonRef',] }],
        cancel: [{ type: Output }],
        submit: [{ type: Output }]
    };
    return VariableGridComponent;
}());
export { VariableGridComponent };
if (false) {
    /** @type {?} */
    VariableGridComponent.prototype.grid;
    /** @type {?} */
    VariableGridComponent.prototype.buttonRef;
    /** @type {?} */
    VariableGridComponent.prototype.cancel;
    /** @type {?} */
    VariableGridComponent.prototype.submit;
    /** @type {?} */
    VariableGridComponent.prototype.data;
    /** @type {?} */
    VariableGridComponent.prototype.columns;
    /**
     * @type {?}
     * @private
     */
    VariableGridComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    VariableGridComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    VariableGridComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    VariableGridComponent.prototype.localePipe;
}
