// @gsp-dip/data-imp-exp/esm5/lib/model/DataImportExportRule.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 导入导出规则
 */
var /**
 * 导入导出规则
 */
DataImportExportRule = /** @class */ (function () {
    function DataImportExportRule() {
        this.ID = '';
        this.Code = '';
        this.Name = '';
        this.Description = '';
        this.Type = 1;
        /**
         * DataExcehangeModel
         */
        this.Models = new Array();
        /**
         * 导入导出规则步骤
         */
        this.Steps = new Array();
        /**
         * 是否导入数据库
         */
        this.ImportToDB = true;
        /**
         * 创建时间
         */
        this.CreatedTime = new Date();
        /**
         * 最后修改时间
         */
        this.LastModifiedTime = new Date();
    }
    return DataImportExportRule;
}());
/**
 * 导入导出规则
 */
export { DataImportExportRule };
if (false) {
    /** @type {?} */
    DataImportExportRule.prototype.ID;
    /** @type {?} */
    DataImportExportRule.prototype.Code;
    /** @type {?} */
    DataImportExportRule.prototype.Name;
    /** @type {?} */
    DataImportExportRule.prototype.Description;
    /** @type {?} */
    DataImportExportRule.prototype.Type;
    /**
     * DataExcehangeModel
     * @type {?}
     */
    DataImportExportRule.prototype.Models;
    /**
     * 导入导出规则步骤
     * @type {?}
     */
    DataImportExportRule.prototype.Steps;
    /**
     * 是否导入数据库
     * @type {?}
     */
    DataImportExportRule.prototype.ImportToDB;
    /**
     * 创建人
     * @type {?}
     */
    DataImportExportRule.prototype.Creator;
    /**
     * 创建时间
     * @type {?}
     */
    DataImportExportRule.prototype.CreatedTime;
    /**
     * 最后修改人
     * @type {?}
     */
    DataImportExportRule.prototype.LastModifier;
    /**
     * 最后修改时间
     * @type {?}
     */
    DataImportExportRule.prototype.LastModifiedTime;
}
