// @gsp-dip/data-imp-exp/esm5/lib/model/IEStepType.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var IEStepType = {
    Default: 0,
    /**
    * 数据导出-数据库抽取
    */
    DataBaseExtract: 10,
    /**
    * 数据导出-数据转换（生成实体）
    */
    ExportMapping: 11,
    /**
    * 数据导出-文件导出
    */
    FileUpdate: 12,
    /**
    * 数据导出-文件下载到客户端
    */
    FileDownload: 13,
    /**
    * 数据导入-文件上传到服务端
    */
    FileUpload: 20,
    /**
    * 数据导入-文件抽取
    */
    FileExtract: 21,
    /**
    * 数据导出-数据转换（生成实体）
    */
    ImportMapping: 22,
    /**
    * 数据导入-构造实体数据
    */
    GenerateData: 23,
    /**
    * 数据导入-数据库更新
    */
    DataBaseUpdate: 24,
    /**
     * 数据校验
     */
    DataCheck: 30,
};
export { IEStepType };
IEStepType[IEStepType.Default] = 'Default';
IEStepType[IEStepType.DataBaseExtract] = 'DataBaseExtract';
IEStepType[IEStepType.ExportMapping] = 'ExportMapping';
IEStepType[IEStepType.FileUpdate] = 'FileUpdate';
IEStepType[IEStepType.FileDownload] = 'FileDownload';
IEStepType[IEStepType.FileUpload] = 'FileUpload';
IEStepType[IEStepType.FileExtract] = 'FileExtract';
IEStepType[IEStepType.ImportMapping] = 'ImportMapping';
IEStepType[IEStepType.GenerateData] = 'GenerateData';
IEStepType[IEStepType.DataBaseUpdate] = 'DataBaseUpdate';
IEStepType[IEStepType.DataCheck] = 'DataCheck';
