// @farris/kendo-binding/esm5/lib/directives/dynamic_validate/farris_dynamic_validate.directive.js
import { Directive, Injector, Input } from "@angular/core";
import { FrameContext } from "@farris/devkit";
var EXCLUDE_VALIDATE_FIELDS = 'excludeValidateRules';
var RULETYPE = 'required';
var FarrisDynamicRequiredDirective = /** @class */ (function () {
    function FarrisDynamicRequiredDirective(frameContext, injector) {
        this.frameContext = frameContext;
        this.injector = injector;
        /**
         * 启用校验的时机
         */
        this.requiredOn = false;
    }
    FarrisDynamicRequiredDirective.prototype.ngOnDestroy = function () {
        this.updateValidateRule(false);
    };
    FarrisDynamicRequiredDirective.prototype.ngOnChanges = function (changes) {
        // 启用必填校验
        this.updateValidateRule(this.requiredOn);
    };
    /**
     *
     * @param isRequired
     */
    FarrisDynamicRequiredDirective.prototype.updateValidateRule = function (isRequired) {
        if (!this.field) {
            return;
        }
        this.frameContext.form.updateFieldValidateRule(this.field, isRequired);
        // if (isRequired) {
        //   const rules = this.frameContext.params.get(EXCLUDE_VALIDATE_FIELDS) || {};
        //   const fieldExcludeRules: string[] = rules[this.field] || [];
        //   if (fieldExcludeRules.includes(RULETYPE)) {
        //     const index = fieldExcludeRules.indexOf(RULETYPE);
        //     fieldExcludeRules.splice(index, 1);
        //   }
        //   this.frameContext.params.set(EXCLUDE_VALIDATE_FIELDS, rules);
        //   this.frameContext.form.updateFieldValidateRule(this.field, fieldExcludeRules);
        // } else {
        //   // 禁用校验
        //   // {int1:[required,maxValue]}
        //   const rules = this.frameContext.params.get(EXCLUDE_VALIDATE_FIELDS) || {};
        //   const fieldExcludeRules = rules[this.field] || [];
        //   if (fieldExcludeRules && !fieldExcludeRules.includes(RULETYPE)) {
        //     fieldExcludeRules.push(RULETYPE);
        //   }
        //   rules[this.field] = fieldExcludeRules;
        //   this.frameContext.params.set(EXCLUDE_VALIDATE_FIELDS, rules);
        //   this.frameContext.form.updateFieldValidateRule(this.field, fieldExcludeRules);
        // }
    };
    FarrisDynamicRequiredDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-dynamic-required]'
                },] }
    ];
    /** @nocollapse */
    FarrisDynamicRequiredDirective.ctorParameters = function () { return [
        { type: FrameContext },
        { type: Injector }
    ]; };
    FarrisDynamicRequiredDirective.propDecorators = {
        field: [{ type: Input, args: ['farris-dynamic-required',] }],
        requiredOn: [{ type: Input, args: ['requiredOn',] }]
    };
    return FarrisDynamicRequiredDirective;
}());
export { FarrisDynamicRequiredDirective };
