// @farris/command-services/esm2015/lib/query-solution/enum/compare-type.enum.js
/**
 * 比较符
 */
export var CompareType;
(function (CompareType) {
    CompareType[CompareType["Equal"] = 0] = "Equal";
    CompareType[CompareType["NotEqual"] = 1] = "NotEqual";
    CompareType[CompareType["Greater"] = 2] = "Greater";
    CompareType[CompareType["GreaterOrEqual"] = 3] = "GreaterOrEqual";
    CompareType[CompareType["Less"] = 4] = "Less";
    CompareType[CompareType["LessOrEqual"] = 5] = "LessOrEqual";
    CompareType[CompareType["Like"] = 6] = "Like";
    CompareType[CompareType["LikeStartWith"] = 7] = "LikeStartWith";
    CompareType[CompareType["LikeEndWith"] = 8] = "LikeEndWith";
    CompareType[CompareType["NotLike"] = 9] = "NotLike";
    CompareType[CompareType["NotLikeStartWith"] = 10] = "NotLikeStartWith";
    CompareType[CompareType["NotLikeEndWith"] = 11] = "NotLikeEndWith";
    CompareType[CompareType["Is"] = 12] = "Is";
    CompareType[CompareType["IsNot"] = 13] = "IsNot";
    CompareType[CompareType["In"] = 14] = "In";
    CompareType[CompareType["NotIn"] = 15] = "NotIn";
})(CompareType || (CompareType = {}));
