// @gsp-wf/wf-process-editor/esm2015/lib/views/shared/event-bus.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { Subject } from 'rxjs/Subject';
import { map, filter } from 'rxjs/operators';
/**
 * @record
 */
export function EventBusMessage() { }
if (false) {
    /** @type {?} */
    EventBusMessage.prototype.channel;
    /** @type {?} */
    EventBusMessage.prototype.payload;
}
export class EventBusService1 {
    constructor() {
        this.listen = new Observable((/**
         * @param {?} observer
         * @return {?}
         */
        observer => {
            this.next = (/**
             * @param {?} payload
             * @return {?}
             */
            (payload) => observer.next(payload));
            return (/**
             * @param {?} _
             * @return {?}
             */
            _ => _ /* nothing to clean */);
        }));
        this.channelEmmiter = this.channelEmmiter.bind(this);
        this.emit = this.emit.bind(this);
    }
    /**
     * @return {?}
     */
    allChannels() {
        return this.listen;
    }
    /**
     * @param {?} channelName
     * @return {?}
     */
    channel(channelName) {
        return this.listen
            .pipe(filter((/**
         * @param {?} e
         * @return {?}
         */
        e => e.channel === channelName)), map((/**
         * @param {?} e
         * @return {?}
         */
        e => e.payload)));
    }
    /**
     * @param {?} channelName
     * @return {?}
     */
    channelEmmiter(channelName) {
        console.log('made listener for', channelName);
        return (/**
         * @param {?} payload
         * @return {?}
         */
        (payload) => {
            this.emit({ channel: channelName, payload });
        });
    }
    /**
     * @param {?} __0
     * @return {?}
     */
    emit({ channel, payload }) {
        console.log('emit', channel);
        this.next({ channel, payload });
    }
}
EventBusService1.decorators = [
    { type: Injectable }
];
/** @nocollapse */
EventBusService1.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    EventBusService1.prototype.listen;
    /**
     * @type {?}
     * @private
     */
    EventBusService1.prototype.next;
}
/**
 * Global EventEmitter
 * @template Events
 */
export class EventBusService {
    constructor() {
        /**
         * Map event name to Subject
         */
        this._subjects = new Map();
    }
    /**
     *
     * @param {?} eventName
     * @param {?} handler
     * @return {?}
     */
    on(eventName, handler) {
        if (!this._subjects.has(eventName)) {
            this._subjects.set(eventName, new Subject());
        }
        /** @type {?} */
        const _subject = this._subjects.get(eventName);
        /** @type {?} */
        const subscription = _subject.subscribe(handler);
        return {
            unsubscribe: (/**
             * @return {?}
             */
            () => {
                subscription.unsubscribe();
                if (_subject.observers.length > 0) {
                    this._subjects.delete(eventName);
                }
            })
        };
    }
    /**
     *
     * @param {?} eventName
     * @param {?} data
     * @return {?}
     */
    emit(eventName, data) {
        if (this._subjects.has(eventName)) {
            this._subjects.get(eventName).next(data);
        }
    }
    /**
     * Destory all subscription and clear the map
     * @return {?}
     */
    destroy() {
        this._subjects.forEach((/**
         * @param {?} subject
         * @param {?} _
         * @return {?}
         */
        (subject, _) => {
            subject.unsubscribe();
        }));
        this._subjects.clear();
    }
}
EventBusService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * Map event name to Subject
     * @type {?}
     * @private
     */
    EventBusService.prototype._subjects;
}
