// @gsp-svc/file-viewer/esm2015/gsp-svc-file-viewer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { ɵg, ɵc, ɵh, ɵi, ɵk, ɵp, ɵq, ɵr, ɵo, ɵn, ɵm, ɵs, ɵl, ɵj, ɵf, ɵb, ɵt, ɵe, ɵz, ɵd, ɵa, ɵv, ɵu, ɵy, FileListModule, FileListComponent, FileViewerService, PdfViewOptions, GspViewerModule, GspViewerComponent, ScrollableDirective } from './public-api';
export { ExcelDocumentComponent as ɵw } from './lib/excel-document/excel-document.component';
export { ExcelPageComponent as ɵx } from './lib/excel-page/excel-page.component';
export { GspTopToolbarComponent as ɵbf } from './lib/gsp-top-toolbar/gsp-top-toolbar.component';
export { GspTopToolbarModule as ɵbe } from './lib/gsp-top-toolbar/gsp-top-toolbar.module';
export { PdfViewerModule as ɵbg } from './lib/pdf-viewer/pdf-viewer.module';
export { PdfViewerComponent as ɵbb } from './lib/pdf-viewer/pdf-viewer/pdf-viewer.component';
export { PptViewerModule as ɵbh } from './lib/ppt-viewer/ppt-viewer.module';
export { PptViewerComponent as ɵbc } from './lib/ppt-viewer/ppt-viewer/ppt-viewer.component';
export { ExcelPageService as ɵba } from './lib/service/excel-page.service';
export { ViewportService as ɵbd } from './lib/service/viewport.service';
