// @gsp-wf/wf-sign/esm5/lib/sign-here/sign-here.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/sign-here/sign-here.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, ElementRef, Input } from '@angular/core';
import { SignatureComponent } from '../signature/signature.component';
var SignHereComponent = /** @class */ (function () {
    function SignHereComponent() {
        // 抛出选中的事件
        this.imageType = "image/png";
        this.imageQuality = 1;
        this.title = '请对审批意见进行签字确认';
        this.isDraw = false;
        this.signaturePadOptions = {
            // passed through to szimek/signature_pad constructor
            'minWidth': 1,
            'canvasWidth': 500,
            'canvasHeight': 240
        };
    }
    /**
     * @return {?}
     */
    SignHereComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    SignHereComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        // this.signaturePad is now available
        this.signaturePad.set('minWidth', 1); // set szimek/signature_pad options at runtime
        this.signaturePad.set('canvasWidth', this.canvasDiv.nativeElement.clientWidth);
        this.signaturePad.clear(); // invoke functions from szimek/signature_pad API
        this.signaturePad.setImageBackground();
        if (this.image) {
            /** @type {?} */
            var img = new Image();
            img.src = this.image;
            this.signaturePad.setBackgroundImage(img);
        }
    };
    /**
     * @return {?}
     */
    SignHereComponent.prototype.confirm = /**
     * @return {?}
     */
    function () {
        console.log(this.signaturePad.toDataURL(this.imageType, this.imageQuality));
        //this.signaturePad.setWaterMark(new Date());
        return this.signaturePad.toDataURL(this.imageType, this.imageQuality);
    };
    /**
     * @return {?}
     */
    SignHereComponent.prototype.reset = /**
     * @return {?}
     */
    function () {
        this.signaturePad.clear();
        this.signaturePad.setImageBackground();
        this.isDraw = false;
    };
    /**
     * @return {?}
     */
    SignHereComponent.prototype.drawComplete = /**
     * @return {?}
     */
    function () {
        this.isDraw = true;
        // will be notified of szimek/signature_pad's onEnd event
        //console.log(this.signaturePad.toDataURL());
    };
    /**
     * @return {?}
     */
    SignHereComponent.prototype.drawStart = /**
     * @return {?}
     */
    function () {
        // will be notified of szimek/signature_pad's onBegin event
        console.log('begin drawing');
    };
    //设置背景图片
    //设置背景图片
    /**
     * @param {?} image
     * @return {?}
     */
    SignHereComponent.prototype.initCanvasByImage = 
    //设置背景图片
    /**
     * @param {?} image
     * @return {?}
     */
    function (image) {
        this.signaturePad.setBackgroundImage(image);
    };
    SignHereComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-sign-here',
                    template: "<div class=\"fm-feature-suggestion fm-feature-section--gutter\">\r\n    <div style=\"padding:2px 2px 2px 0;text-align:left;width:100%;height: 21px;\">\r\n        <span style=\"font-size: 15px;color: #999999;\">{{title}}</span>\r\n      </div>\r\n      <div class=\"wf-sign-signature-docker\">\r\n      <div #canvastitle class=\"wf-sign-signature\">\r\n      <wf-sign-signature  [options]=\"signaturePadOptions\" (onBeginEvent)=\"drawStart()\" (onEndEvent)=\"drawComplete()\"></wf-sign-signature>\r\n    </div>\r\n</div>\r\n</div>\r\n",
                    styles: [".fm-page-main1{overflow:hidden}.fm-utils-flex-column{display:flex;flex-direction:column}.fm-page{position:absolute;top:0;left:0;bottom:0;right:0;display:flex;flex-direction:column;background:#fff;font-size:13px;line-height:18px}.fm-page-main{flex-basis:0;flex-shrink:1;flex-grow:1;overflow-x:hidden;overflow-y:auto;position:relative}.fm-feature-section--gutter{margin:0 0 12px}.fm-feature-suggestion{background:#fff;padding:12px 16px}.fm-feature-suggestion .fm-feature-suggestion--bar{display:flex;flex-direction:row;align-items:center;color:#888}.wf-sign-signature{margin:1px}.wf-sign-signature-docker{border:1px dashed #ddd;margin-left:4px;margin-right:4px}"]
                }] }
    ];
    /** @nocollapse */
    SignHereComponent.ctorParameters = function () { return []; };
    SignHereComponent.propDecorators = {
        signaturePad: [{ type: ViewChild, args: [SignatureComponent,] }],
        canvasDiv: [{ type: ViewChild, args: ['canvastitle',] }],
        imageType: [{ type: Input }],
        imageQuality: [{ type: Input }],
        title: [{ type: Input }],
        image: [{ type: Input }]
    };
    return SignHereComponent;
}());
export { SignHereComponent };
if (false) {
    /** @type {?} */
    SignHereComponent.prototype.signaturePad;
    /** @type {?} */
    SignHereComponent.prototype.canvasDiv;
    /** @type {?} */
    SignHereComponent.prototype.imageType;
    /** @type {?} */
    SignHereComponent.prototype.imageQuality;
    /** @type {?} */
    SignHereComponent.prototype.title;
    /** @type {?} */
    SignHereComponent.prototype.image;
    /** @type {?} */
    SignHereComponent.prototype.isDraw;
    /** @type {?} */
    SignHereComponent.prototype.signaturePadOptions;
}
