// @farris/ui-datagrid-settings/esm2015/lib/multi-headers/new-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, ElementRef } from '@angular/core';
export class NewHeaderFormComponent {
    constructor() {
        this.label = '表头名称';
        this.placeholder = "输入表头名称";
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
}
NewHeaderFormComponent.decorators = [
    { type: Component, args: [{
                selector: 'group-header-form',
                template: `
    <div class="d-flex flex-row" style="align-items: center;height: 100%;">
        <span style="width: 90px;text-align: right;">{{label}}</span>
        <input #txt placeholder="{{placeholder}}" type="text" [(ngModel)]="title" class="form-control ml-2" style="width: 210px">
    </div>
    `
            }] }
];
/** @nocollapse */
NewHeaderFormComponent.ctorParameters = () => [];
NewHeaderFormComponent.propDecorators = {
    label: [{ type: Input }],
    title: [{ type: Input }],
    placeholder: [{ type: Input }],
    input: [{ type: ViewChild, args: ['txt',] }]
};
if (false) {
    /** @type {?} */
    NewHeaderFormComponent.prototype.label;
    /** @type {?} */
    NewHeaderFormComponent.prototype.title;
    /** @type {?} */
    NewHeaderFormComponent.prototype.placeholder;
    /** @type {?} */
    NewHeaderFormComponent.prototype.input;
}
