// @farris/ui-datagrid-settings/esm2015/lib/multi-headers/setcolumns.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from "@angular/core";
export class SetColumnsComponent {
    constructor() {
        this.data = [];
        this.checkedItems = [];
        this.selectedIds = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.selectedIds = (this.checkedItems || []).map((/**
         * @param {?} n
         * @return {?}
         */
        n => n.id));
        if (this.data) {
            this.data.forEach((/**
             * @param {?} item
             * @return {?}
             */
            item => {
                if (!item.id) {
                    item.id = item.field;
                }
            }));
        }
    }
}
SetColumnsComponent.decorators = [
    { type: Component, args: [{
                selector: 'set-columns',
                template: "<multi-select [dataSource]=\"data\"\r\n[idField]=\"'id'\" [valueField]=\"'field'\" [textField]=\"'title'\" [displayType]=\"'List'\"\r\n[enableTargetSort]=\"true\" [showCheckbox]=\"true\" [enableMoveMore]=\"true\" [(selectedId)]=\"selectedIds\">\r\n</multi-select>"
            }] }
];
/** @nocollapse */
SetColumnsComponent.ctorParameters = () => [];
SetColumnsComponent.propDecorators = {
    data: [{ type: Input }],
    checkedItems: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    SetColumnsComponent.prototype.data;
    /** @type {?} */
    SetColumnsComponent.prototype.checkedItems;
    /** @type {?} */
    SetColumnsComponent.prototype.selectedIds;
}
