// @gsp-wf/biz-process-api/bundles/gsp-wf-biz-process-api.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@gsp-wf/biz-process-api', ['exports', '@angular/core'], factory) :
    (factory((global['gsp-wf'] = global['gsp-wf'] || {}, global['gsp-wf']['biz-process-api'] = {}),global.ng.core));
}(this, (function (exports,i0) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BizProcessApiService = /** @class */ (function () {
        function BizProcessApiService() {
        }
        BizProcessApiService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */
        BizProcessApiService.ctorParameters = function () { return []; };
        /** @nocollapse */ BizProcessApiService.ngInjectableDef = i0.defineInjectable({ factory: function BizProcessApiService_Factory() { return new BizProcessApiService(); }, token: BizProcessApiService, providedIn: "root" });
        return BizProcessApiService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BizProcessApiComponent = /** @class */ (function () {
        function BizProcessApiComponent() {
        }
        /**
         * @return {?}
         */
        BizProcessApiComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
            };
        BizProcessApiComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'lib-biz-process-api',
                        template: "\n    <p>\n      biz-process-api works!\n    </p>\n  "
                    }] }
        ];
        /** @nocollapse */
        BizProcessApiComponent.ctorParameters = function () { return []; };
        return BizProcessApiComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BizProcessApiModule = /** @class */ (function () {
        function BizProcessApiModule() {
        }
        BizProcessApiModule.decorators = [
            { type: i0.NgModule, args: [{
                        declarations: [BizProcessApiComponent],
                        imports: [],
                        exports: [BizProcessApiComponent]
                    },] }
        ];
        return BizProcessApiModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.BizProcessApiService = BizProcessApiService;
    exports.BizProcessApiComponent = BizProcessApiComponent;
    exports.BizProcessApiModule = BizProcessApiModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
