// @farris/ui-datagrid/esm2015/lib/components/filter-toolbar/smart-filter-toolbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
import { Component, EventEmitter, Input, Output, ElementRef, Renderer2, Injector, ViewChild } from '@angular/core';
import { DatagridComponent } from '../../datagrid.component';
import { takeUntil } from 'rxjs/operators';
export class SmartFilterToolbarComponent {
    /**
     * @param {?} el
     * @param {?} render
     * @param {?} inject
     */
    constructor(el, render, inject) {
        this.el = el;
        this.render = render;
        this.inject = inject;
        this.padding = '';
        this.height = 40;
        this.filterData = [];
        this.clearFilters = new EventEmitter();
        this.removeFilter = new EventEmitter();
        this.filterChange = new EventEmitter();
        this.resize = new EventEmitter();
        this.sk = '';
        this.destory$ = new Subject();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.dg = this.inject.get(DatagridComponent);
        this.sk = 'smartfilter_' + this.dg.dgs.createConfigKey(this.dg.id);
        this.dg.dgs.clearFilter.pipe(takeUntil(this.destory$)).subscribe((/**
         * @return {?}
         */
        () => {
            this.clearFilters.emit(null);
        }));
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.destory$.next();
        this.destory$.complete();
    }
    /**
     * 条件变化事件
     * @param {?} $event
     * @return {?}
     */
    onQuery($event) {
        const { data, filterItem } = $event;
        if (data && data.length) {
            /** @type {?} */
            const conditions = JSON.parse(data);
            if (conditions && conditions.length) {
                this.filterChange.emit({ conditions, filterItem });
            }
        }
    }
    /**
     * 清除所有条件
     * @param {?} $event
     * @return {?}
     */
    onResetFilter($event) {
        this.clearFilters.emit($event);
    }
    /**
     * 移除单个条件
     * @param {?} $event
     * @return {?}
     */
    onRemoveFilter($event) {
        this.removeFilter.emit($event);
    }
    /**
     * 展开折叠事件
     * @param {?} $event
     * @return {?}
     */
    onResize($event) {
        setTimeout((/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const smartFilterWrapper = this.el.nativeElement.querySelector('.f-filter-wrapper');
            if (smartFilterWrapper) {
                /** @type {?} */
                const h = smartFilterWrapper.offsetHeight;
                if (h) {
                    this.resize.emit(h + 6);
                }
            }
        }));
    }
}
SmartFilterToolbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'smart-filter-toolbar',
                template: "\r\n<div class=\"f-datagrid-filter-toolbar\" [style.height.px]=\"height\" \r\n[style.padding]=\"padding\" placeholder=\"\u6B64\u5904\u663E\u793A\u5217\u8FC7\u6EE4\u6761\u4EF6\" style=\"align-items: normal;\"\r\n[style.display]=\"filterData && filterData.length ? '' : 'none'\">\r\n<farris-filter [filterList]=\"filterData\" (query)=\"onQuery($event)\"  #farrisFilter\r\n    (resetChange)=\"onResetFilter($event)\" (clearFilter)=\"onRemoveFilter($event)\"\r\n    (filterResize)=\"onResize($event)\" (expandFilter)=\"onResize($event)\"\r\n    [localStorageKey]=\"sk\"\r\n>\r\n</farris-filter>\r\n\r\n</div>"
            }] }
];
/** @nocollapse */
SmartFilterToolbarComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: Injector }
];
SmartFilterToolbarComponent.propDecorators = {
    padding: [{ type: Input }],
    height: [{ type: Input }],
    filterData: [{ type: Input }],
    clearFilters: [{ type: Output }],
    removeFilter: [{ type: Output }],
    filterChange: [{ type: Output }],
    resize: [{ type: Output }],
    filtebar: [{ type: ViewChild, args: ['farrisFilter',] }]
};
if (false) {
    /** @type {?} */
    SmartFilterToolbarComponent.prototype.padding;
    /** @type {?} */
    SmartFilterToolbarComponent.prototype.height;
    /** @type {?} */
    SmartFilterToolbarComponent.prototype.filterData;
    /** @type {?} */
    SmartFilterToolbarComponent.prototype.clearFilters;
    /** @type {?} */
    SmartFilterToolbarComponent.prototype.removeFilter;
    /** @type {?} */
    SmartFilterToolbarComponent.prototype.filterChange;
    /** @type {?} */
    SmartFilterToolbarComponent.prototype.resize;
    /** @type {?} */
    SmartFilterToolbarComponent.prototype.filtebar;
    /** @type {?} */
    SmartFilterToolbarComponent.prototype.sk;
    /** @type {?} */
    SmartFilterToolbarComponent.prototype.dg;
    /** @type {?} */
    SmartFilterToolbarComponent.prototype.destory$;
    /**
     * @type {?}
     * @private
     */
    SmartFilterToolbarComponent.prototype.el;
    /**
     * @type {?}
     * @private
     */
    SmartFilterToolbarComponent.prototype.render;
    /**
     * @type {?}
     * @private
     */
    SmartFilterToolbarComponent.prototype.inject;
}
