// @farris/mobile-devkit/esm5/lib/entity/entity-path/data_path.js
/*
 * @Author: Witt
 * @Date: 2018-12-27 09:26:41
 * @Last Modified by: Witt
 * @Last Modified time: 2019-01-15 22:00:45
 */
import { DataPathNode } from './data_path_node';
/**
 * 变更路径（简单双向列表）
 */
var DataPath = /** @class */ (function () {
    /**
     * 构造函数
     */
    function DataPath() {
        this.head = new DataPathNode(null, null);
        this.length = 0;
    }
    /**
     * 添加一个节点到头部
     */
    DataPath.prototype.unshift = function (type, data) {
        var newNode = new DataPathNode(type, data);
        newNode.next = this.head.next;
        newNode.prev = this.head;
        this.head.next = newNode;
        if (newNode.next) {
            newNode.next.prev = newNode;
        }
        this.length++;
    };
    /**
     * 在链表最后追加一个节点
     */
    DataPath.prototype.push = function (type, data) {
        var tailNode = this.getTail();
        var newNode = new DataPathNode(type, data);
        tailNode.next = newNode;
        this.length++;
    };
    /**
     * 获取链表尾部节点
     */
    DataPath.prototype.getTail = function () {
        var lastNode = this.head;
        while (lastNode.next) {
            lastNode = lastNode.next;
        }
        return lastNode;
    };
    /**
     * 转换为数组格式
     */
    DataPath.prototype.toArray = function () {
        var pathArray = [];
        var currentNode = this.head.next;
        while (currentNode) {
            pathArray.push(currentNode.type + ":" + currentNode.value);
            currentNode = currentNode.next;
        }
        return pathArray;
    };
    /**
     * 转换为字符串格式
     */
    DataPath.prototype.toString = function () {
        var pathArray = this.toArray();
        var pathString = pathArray.join(', ');
        return "[" + pathString + "]";
    };
    /**
     * 拷贝
     */
    DataPath.prototype.clone = function () {
        var newDataPath = new DataPath();
        var curDataNode = this.head.next;
        while (curDataNode) {
            newDataPath.push(curDataNode.type, curDataNode.value);
            curDataNode = curDataNode.next;
        }
        return newDataPath;
    };
    return DataPath;
}());
export { DataPath };
