// @gsp-sys/sysmgr-common/esm2015/lib/config/sys-security-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { SysCommonUtilService } from '../service/sys-manager-util.service';
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../service/sys-manager-util.service";
/**
 * 保密测评服务类
 */
export class SysSecurityConfigService {
    /**
     * @param {?} http
     * @param {?} sysUtil
     */
    constructor(http, sysUtil) {
        this.http = http;
        this.sysUtil = sysUtil;
    }
    /**
     * 是否开启三员
     * @return {?}
     */
    isOnSecurity() {
        /** @type {?} */
        const url = this.sysUtil.sysSecurityApi + '/useSecurity';
        return this.http.get(url, this.sysUtil.getHeader()).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        (res) => (/** @type {?} */ (res)))));
    }
    /**
     * 根据当前登录用户获取保密测评的功能配置范围
     * @param {?} menuId 功能标识
     * @return {?}
     */
    getCurrentUserAppConfig(menuId) {
        /** @type {?} */
        const url = this.sysUtil.sysSecurityApi + '/funcobj/' + menuId;
        return this.http.get(url, this.sysUtil.getHeader());
    }
    /**
     * 根据指定用户类别获取界面显示信息-,目前只传递type="alluser"使用
     * @param {?} menuId 功能标识
     * @param {?} type
     * @return {?}
     */
    getAppConfigByUserType(menuId, type) {
        /** @type {?} */
        const url = this.sysUtil.sysSecurityApi + '/funcobj/' + menuId + '/' + type;
        return this.http.get(url, this.sysUtil.getHeader());
    }
}
SysSecurityConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
SysSecurityConfigService.ctorParameters = () => [
    { type: HttpClient },
    { type: SysCommonUtilService }
];
/** @nocollapse */ SysSecurityConfigService.ngInjectableDef = i0.defineInjectable({ factory: function SysSecurityConfigService_Factory() { return new SysSecurityConfigService(i0.inject(i1.HttpClient), i0.inject(i2.SysCommonUtilService)); }, token: SysSecurityConfigService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    SysSecurityConfigService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    SysSecurityConfigService.prototype.sysUtil;
}
