// @farris/extend-header-footer/esm5/lib/components/page-header/page-header.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/page-header/page-header.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter, ContentChild } from '@angular/core';
import { Observable, Subject } from 'rxjs';
import { FEPageHeaderDefault } from '../../models/page-header.model';
import { FEPageHeaderContentDirective, FEPageHeaderDownDirective, FEPageHeaderTitleDirective } from '../../directives/index';
import { FExtendUtils } from '../../models/utils';
var FEPageHeaderComponent = /** @class */ (function () {
    function FEPageHeaderComponent() {
        this._settings = Object.assign({}, FEPageHeaderDefault);
        this.realToolbarDatas = [];
        // 展示类型
        this.showType = "modal";
        this.btnDisable = new Subject();
        this.btnVisible = new Subject();
        // 事件
        this.pgPreClickEvent = new EventEmitter();
        this.pgNextClickEvent = new EventEmitter();
        this.toolbarClickEvent = new EventEmitter();
        this.viewChangeEvent = new EventEmitter();
    }
    Object.defineProperty(FEPageHeaderComponent.prototype, "options", {
        get: /**
         * @return {?}
         */
        function () {
            return this._settings;
        },
        set: /**
         * @param {?} settings
         * @return {?}
         */
        function (settings) {
            if (settings) {
                this._settings = Object.assign({}, FEPageHeaderDefault, settings);
                this.realToolbarDatas = this._settings['toolbarDatas'];
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FEPageHeaderComponent.prototype, "toolbarBtnDisable", {
        // 按钮的禁用状态
        set: 
        // 按钮的禁用状态
        /**
         * @param {?} values
         * @return {?}
         */
        function (values) {
            this.btnDisable = values;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FEPageHeaderComponent.prototype, "toolbarBtnVisible", {
        // 按钮的查看状态
        set: 
        // 按钮的查看状态
        /**
         * @param {?} values
         * @return {?}
         */
        function (values) {
            this.btnVisible = values;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FEPageHeaderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    // 变化
    // 变化
    /**
     * @param {?} changes
     * @return {?}
     */
    FEPageHeaderComponent.prototype.ngOnChanges = 
    // 变化
    /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.hasOwnProperty('toolbarAllDatas') || changes.hasOwnProperty('toolbarConfig') || changes.hasOwnProperty('showType')) {
            if (this.toolbarKey && this.toolbarConfig && this.toolbarAllDatas) {
                this.realToolbarDatas = tslib_1.__spread(FExtendUtils.formToolbarData(this.toolbarAllDatas, this.toolbarConfig, this.toolbarKey, this.showType));
            }
        }
    };
    // 向上翻页
    // 向上翻页
    /**
     * @param {?} ev
     * @return {?}
     */
    FEPageHeaderComponent.prototype.pgPreClickHandler = 
    // 向上翻页
    /**
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        if (!this.options['pgPreDisabled']) {
            this.pgPreClickEvent.emit(ev);
        }
    };
    // 向下翻页
    // 向下翻页
    /**
     * @param {?} ev
     * @return {?}
     */
    FEPageHeaderComponent.prototype.pgNextClickHandler = 
    // 向下翻页
    /**
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        if (!this.options['pgNextDisabled']) {
            this.pgNextClickEvent.emit(ev);
        }
    };
    // 抛出事件
    // 抛出事件
    /**
     * @param {?} ev
     * @return {?}
     */
    FEPageHeaderComponent.prototype.responseToolbarClickHandler = 
    // 抛出事件
    /**
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        this.toolbarClickEvent.emit(ev);
    };
    // View改变
    // View改变
    /**
     * @param {?} ev
     * @return {?}
     */
    FEPageHeaderComponent.prototype.toolTypeChangeHandler = 
    // View改变
    /**
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        this.viewChangeEvent.emit(ev);
    };
    FEPageHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fe-page-header',
                    template: "\n  <div class=\"f-page-header\"  [ngClass]=\"'showtype-'+showType\">\n  <nav class=\"f-page-header-base\">\n      <ng-container *ngIf=\"hTitleDirective &&hTitleDirective.templateRef\">\n          <div class=\"f-title\" [ngClass]=\"hTitleDirective.cls\">\n              <ng-template [ngTemplateOutlet]=\"hTitleDirective.templateRef\"></ng-template>\n          </div>\n      </ng-container>\n      <div class=\"f-title\" *ngIf=\"!hTitleDirective\">\n          <span *ngIf=\"options['showIcon']\" class=\"f-title-icon\" [ngClass]=\"options['iconCls']\">\n              <i class=\"f-icon\" [ngClass]=\"options['iconName']\"></i>\n          </span>\n          <h4 class=\"f-title-text\" [innerHTML]=\"options['mainTitle']|feSafeHtml\"></h4>\n          <h5 *ngIf=\"options['subTitle']\" class=\"f-title-subtitle\" [innerHTML]=\"options['subTitle']|feSafeHtml\"></h5>\n          <div *ngIf=\"options['status']\" class=\"f-title-status\" [innerHTML]=\"options['status']|feSafeHtml\">\n          </div>\n          <div class=\"f-title-pagination\" *ngIf=\"options['pgShow']\">\n              <span class=\"f-icon f-icon-arrow-w\" [ngClass]=\"{'f-state-disabled':options['pgPreDisabled']}\"\n                  (click)=\"pgPreClickHandler($event)\"></span>\n              <span class=\"f-icon f-icon-arrow-e\" [ngClass]=\"{'f-state-disabled':options['pgNextDisabled']}\"\n                  (click)=\"pgNextClickHandler($event)\"></span>\n          </div>\n      </div>\n      <ng-container *ngIf=\"hContentDirective &&hContentDirective.templateRef\">\n          <div class=\"f-content\" [ngClass]=\"hContentDirective.cls\">\n              <ng-template [ngTemplateOutlet]=\"hContentDirective.templateRef\"></ng-template>\n          </div>\n      </ng-container>\n      <ng-container *ngIf=\"options['content']&&!hContentDirective\" >\n          <div class=\"f-content\"  [innerHTML]=\"options['content']|feSafeHtml\"></div>\n      </ng-container>\n      <ng-container *ngIf=\"realToolbarDatas.length>0\">\n          <f-response-toolbar class=\"f-toolbar ml-auto\" [ngClass]=\"options['toolbarCls']\" [datas]=\"realToolbarDatas\"\n          [btnState]=\"btnDisable\" [btnVisible]=\"btnVisible\"  [buttonSize]=\"options['toolbarBtnSize']=='default'?'':options['toolbarBtnSize']\" [popDirection]=\"options['toolbarPopDirection']=='default'?'':options['toolbarPopDirection']\"\n              (rtClickEvent)=\"responseToolbarClickHandler($event)\">\n          </f-response-toolbar>\n      </ng-container>\n      <ng-container *ngIf=\"options['showViewChange']&&options['viewDatas'].length\">\n          <farris-view-change [toolbarData]=\"options['viewDatas']\" class=\"f-btn-ml\" [viewType]=\"options['viewType']\"\n              [defaultType]=\"options['viewDefaultType']\" [viewGroupId]=\"options['viewGroupId']\"\n              (toolTypeChange)=\"toolTypeChangeHandler($event)\"></farris-view-change>\n      </ng-container>\n  </nav>\n  <ng-container *ngIf=\"hExtendDirective &&  hExtendDirective.templateRef\">\n      <div class=\"f-content\" [ngClass]=\"hExtendDirective.cls\">\n          <ng-template [ngTemplateOutlet]=\"hExtendDirective.templateRef\"></ng-template>\n      </div>\n  </ng-container>\n  <div class=\"f-page-header-extend\" *ngIf=\"options['extend']&&!hExtendDirective\"  [innerHTML]=\"options['extend']|feSafeHtml\">\n  </div>\n</div>\n  "
                }] }
    ];
    /** @nocollapse */
    FEPageHeaderComponent.ctorParameters = function () { return []; };
    FEPageHeaderComponent.propDecorators = {
        options: [{ type: Input }],
        toolbarAllDatas: [{ type: Input }],
        toolbarConfig: [{ type: Input }],
        toolbarKey: [{ type: Input }],
        showType: [{ type: Input }],
        hContentDirective: [{ type: ContentChild, args: [FEPageHeaderContentDirective,] }],
        hExtendDirective: [{ type: ContentChild, args: [FEPageHeaderDownDirective,] }],
        hTitleDirective: [{ type: ContentChild, args: [FEPageHeaderTitleDirective,] }],
        toolbarBtnDisable: [{ type: Input }],
        toolbarBtnVisible: [{ type: Input }],
        pgPreClickEvent: [{ type: Output }],
        pgNextClickEvent: [{ type: Output }],
        toolbarClickEvent: [{ type: Output }],
        viewChangeEvent: [{ type: Output }]
    };
    return FEPageHeaderComponent;
}());
export { FEPageHeaderComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FEPageHeaderComponent.prototype._settings;
    /** @type {?} */
    FEPageHeaderComponent.prototype.realToolbarDatas;
    /** @type {?} */
    FEPageHeaderComponent.prototype.toolbarAllDatas;
    /** @type {?} */
    FEPageHeaderComponent.prototype.toolbarConfig;
    /** @type {?} */
    FEPageHeaderComponent.prototype.toolbarKey;
    /** @type {?} */
    FEPageHeaderComponent.prototype.showType;
    /** @type {?} */
    FEPageHeaderComponent.prototype.hContentDirective;
    /** @type {?} */
    FEPageHeaderComponent.prototype.hExtendDirective;
    /** @type {?} */
    FEPageHeaderComponent.prototype.hTitleDirective;
    /** @type {?} */
    FEPageHeaderComponent.prototype.btnDisable;
    /** @type {?} */
    FEPageHeaderComponent.prototype.btnVisible;
    /** @type {?} */
    FEPageHeaderComponent.prototype.pgPreClickEvent;
    /** @type {?} */
    FEPageHeaderComponent.prototype.pgNextClickEvent;
    /** @type {?} */
    FEPageHeaderComponent.prototype.toolbarClickEvent;
    /** @type {?} */
    FEPageHeaderComponent.prototype.viewChangeEvent;
}
