// @gsp-cmp/coderule-selector/esm5/lib/code-rule-select.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/code-rule-select.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { TreeViewModule } from '@progress/kendo-angular-treeview';
import { GridModule } from '@progress/kendo-angular-grid';
// import { FarrisDialogModule } from '@farris/ui-dialog';
import { SelectCodeRuleComponent } from './components/select-code-rule/select-code-rule.component';
import { CodeRuleListComponent } from './components/code-rule-list/code-rule-list.component';
import { CodeRuleSelectService } from './service/code-rule-select.service';
var CodeRuleSelectModule = /** @class */ (function () {
    function CodeRuleSelectModule() {
    }
    CodeRuleSelectModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        TreeViewModule,
                        GridModule,
                    ],
                    exports: [
                        SelectCodeRuleComponent
                    ],
                    providers: [CodeRuleSelectService],
                    // 服务提供者
                    declarations: [SelectCodeRuleComponent, CodeRuleListComponent],
                    entryComponents: [SelectCodeRuleComponent]
                },] }
    ];
    /** @nocollapse */
    CodeRuleSelectModule.ctorParameters = function () { return []; };
    return CodeRuleSelectModule;
}());
export { CodeRuleSelectModule };
