// @gsp-svc/formdoc-upload/esm2015/lib/upload/upload.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { LOCALE_ID, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FileUploadComponent } from './file-upload/file-upload.component';
import { NotifyModule } from '@farris/ui-notify';
import { MessagerModule } from '@farris/ui-messager';
import { UploadService } from './upload.service';
import { LoadingModule } from '@farris/ui-loading';
import { ProgressModule } from '@farris/ui-progress';
import { FormsModule } from '@angular/forms';
import { ComboListModule } from '@farris/ui-combo-list';
import { FarrisFormsModule } from '@farris/ui-forms';
import { DownloadService } from './download.service';
import { LocalLangPipe } from '../local.pipe';
import { CacheService, HttpService, LanguageService, SessionService } from '@ecp-caf/caf-common';
import { HttpClientModule } from '@angular/common/http';
const ɵ0 = localStorage.getItem('languageCode') || 'zh-CHS';
export class UploadModule {
}
UploadModule.decorators = [
    { type: NgModule, args: [{
                declarations: [FileUploadComponent, LocalLangPipe],
                imports: [
                    CommonModule,
                    NotifyModule,
                    MessagerModule.forRoot(),
                    LoadingModule.forRoot(),
                    ProgressModule,
                    FormsModule,
                    FarrisFormsModule,
                    ComboListModule,
                    HttpClientModule
                ],
                exports: [FileUploadComponent],
                entryComponents: [FileUploadComponent],
                providers: [UploadService, DownloadService, LanguageService, HttpService, CacheService, SessionService,
                    { provide: LOCALE_ID, useValue: ɵ0 },]
            },] }
];
/** @type {?} */
export const ɵa = UploadModule;
export { ɵ0 };
