// @farris/ide-devkit/esm5/lib/components/ng2tree-common/models/tree-node.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter } from '@angular/core';
import { TREE_EVENTS } from '../constants/events';
var TreeNode = /** @class */ (function () {
    function TreeNode(data, parent, treeModel) {
        if (parent === void 0) { parent = null; }
        var _this = this;
        this.data = data;
        this.isVirtualRoot = false;
        this._isActive = false;
        // 变更树节点展示字段，由外部触发
        this.changeDisplayField = new EventEmitter();
        Object.assign(this, data, { parent: parent, treeModel: treeModel });
        this.treeModel = treeModel;
        this.parent = parent;
        this.level = this.parent ? this.parent.level + 1 : 0;
        if (this.isExpandedField) {
            this.isExpanded = true;
        }
        // 验证字段添加
        // console.log(this.idField);
        // console.log(this.typeField);
        this.childrenField = this.childrenField.map(function (child) { return new TreeNode(child, _this, treeModel); });
        // bug修复：TreeModel.canMoveNode使用parentNode，会出现undefined。这里做兼容
        this.parentNode = this.parent;
    }
    Object.defineProperty(TreeNode.prototype, "isExpanded", {
        get: /**
         * @return {?}
         */
        function () { return this._isExpanded; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this._isExpanded = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeNode.prototype, "isCollapsed", {
        get: /**
         * @return {?}
         */
        function () { return !this._isExpanded; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeNode.prototype, "isActive", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isActive;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._isActive = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeNode.prototype, "options", {
        // Proxy of treeModel
        get: 
        // Proxy of treeModel
        /**
         * @return {?}
         */
        function () {
            return this.treeModel.options;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    TreeNode.prototype.fireEvent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.treeModel.fireEvent(event);
    };
    Object.defineProperty(TreeNode.prototype, "idField", {
        // 表现的更像一个字段，将其封装成属性，并根据开放的读写权限，设置get/set
        get: 
        // 表现的更像一个字段，将其封装成属性，并根据开放的读写权限，设置get/set
        /**
         * @return {?}
         */
        function () {
            return this[this.options.idField];
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this[this.options.idField] = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeNode.prototype, "childrenField", {
        get: /**
         * @return {?}
         */
        function () {
            return this[this.options.childrenField] || [];
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this[this.options.childrenField] = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeNode.prototype, "displayField", {
        get: /**
         * @return {?}
         */
        function () {
            return this[this.options.displayField];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeNode.prototype, "typeField", {
        get: /**
         * @return {?}
         */
        function () {
            return this[this.options.typeField];
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this[this.options.typeField] = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeNode.prototype, "isExpandedField", {
        get: /**
         * @return {?}
         */
        function () {
            return this[this.options.isExpandedField];
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this[this.options.isExpandedField] = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TreeNode.prototype.allowDrag = /**
     * @return {?}
     */
    function () {
        return this.options.allowDrag;
    };
    /**
     * @return {?}
     */
    TreeNode.prototype.enableCustomContextMenu = /**
     * @return {?}
     */
    function () {
        return this.options.enableCustomContextMenu;
    };
    Object.defineProperty(TreeNode.prototype, "isRoot", {
        get: /**
         * @return {?}
         */
        function () { return this.parent.isVirtualRoot; } // 父节点为空，表示这个一个根节点
        ,
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeNode.prototype, "realParent", {
        get: 
        // 父节点为空，表示这个一个根节点
        /**
         * @return {?}
         */
        function () { return this.isRoot ? null : this.parent; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeNode.prototype, "isLeaf", {
        get: /**
         * @return {?}
         */
        function () { return !this.childrenField.length; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeNode.prototype, "hasChildren", {
        get: /**
         * @return {?}
         */
        function () { return !this.isLeaf; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TreeNode.prototype, "isFocused", {
        get: /**
         * @return {?}
         */
        function () { return this.treeModel.focusedNode == this; },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TreeNode.prototype.getFirstChild = /**
     * @return {?}
     */
    function () {
        return this.childrenField && this.childrenField[0];
    };
    /**
     * @return {?}
     */
    TreeNode.prototype.getLastChild = /**
     * @return {?}
     */
    function () {
        return this.childrenField && this.childrenField[this.childrenField.length - 1];
    };
    /**
     * @private
     * @return {?}
     */
    TreeNode.prototype._getIndexInParent = /**
     * @private
     * @return {?}
     */
    function () {
        return this.parent && this.parent.childrenField.indexOf(this);
    };
    /**
     * @param {?} steps
     * @return {?}
     */
    TreeNode.prototype.findAdjacentSibling = /**
     * @param {?} steps
     * @return {?}
     */
    function (steps) {
        /** @type {?} */
        var index = this._getIndexInParent();
        return this.parent && this.parent.childrenField[index + steps];
    };
    /**
     * @return {?}
     */
    TreeNode.prototype.findNextSibling = /**
     * @return {?}
     */
    function () {
        return this.findAdjacentSibling(+1);
    };
    /**
     * @return {?}
     */
    TreeNode.prototype.findPreviousSibling = /**
     * @return {?}
     */
    function () {
        return this.findAdjacentSibling(-1);
    };
    /**
     * @param {?=} goInside
     * @return {?}
     */
    TreeNode.prototype.findNextNode = /**
     * @param {?=} goInside
     * @return {?}
     */
    function (goInside) {
        if (goInside === void 0) { goInside = true; }
        return goInside && this.isExpanded && this.getFirstChild() ||
            this.findNextSibling() ||
            this.parent && this.parent.findNextNode(false);
    };
    /**
     * @return {?}
     */
    TreeNode.prototype.findPreviousNode = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var previousSibling = this.findPreviousSibling();
        if (!previousSibling) {
            return this.realParent;
        }
        return previousSibling.isCollapsed
            ? previousSibling
            : previousSibling.getLastChild();
    };
    /**
     * @param {?} node
     * @return {?}
     */
    TreeNode.prototype.isDescendantOf = /**
     * @param {?} node
     * @return {?}
     */
    function (node) {
        if (this === node) {
            return true;
        }
        else {
            return this.parent && this.parent.isDescendantOf(node);
        }
    };
    // 切换方法
    // 切换节点的折叠（Collapsed）与扩展（Expanded）状态
    // 切换方法
    // 切换节点的折叠（Collapsed）与扩展（Expanded）状态
    /**
     * @return {?}
     */
    TreeNode.prototype.toggle = 
    // 切换方法
    // 切换节点的折叠（Collapsed）与扩展（Expanded）状态
    /**
     * @return {?}
     */
    function () {
        this.isExpanded = !this.isExpanded;
        this.fireEvent({ eventName: TREE_EVENTS.onToggle, node: this, isExpanded: this.isExpanded });
    };
    /**
     * @private
     * @return {?}
     */
    TreeNode.prototype._activate = /**
     * @private
     * @return {?}
     */
    function () {
        this._isActive = true;
        this.fireEvent({ eventName: TREE_EVENTS.onActivate, node: this });
        this.focus();
    };
    /**
     * @private
     * @return {?}
     */
    TreeNode.prototype._deactivate = /**
     * @private
     * @return {?}
     */
    function () {
        this._isActive = false;
        this.fireEvent({ eventName: TREE_EVENTS.onDeactivate, node: this });
    };
    /**
     * @return {?}
     */
    TreeNode.prototype.toggleActivated = /**
     * @return {?}
     */
    function () {
        if (this.isActive) {
            this._deactivate();
            this.treeModel.activeNode = null;
        }
        else {
            if (this.treeModel.activeNode) {
                this.treeModel.activeNode._deactivate();
            }
            this._activate();
            this.treeModel.activeNode = this;
        }
        this.fireEvent({ eventName: TREE_EVENTS.onActiveChanged, node: this, isActive: this.isActive });
    };
    /**
     * @return {?}
     */
    TreeNode.prototype.focus = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var previousNode = this.treeModel.focusedNode;
        this.treeModel.focusedNode = this;
        if (previousNode) {
            this.fireEvent({ eventName: TREE_EVENTS.onBlur, node: previousNode });
        }
        this.fireEvent({ eventName: TREE_EVENTS.onFocus, node: this });
    };
    /**
     * @return {?}
     */
    TreeNode.prototype.blur = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var previousNode = this.treeModel.focusedNode;
        this.treeModel.focusedNode = null;
        if (previousNode) {
            this.fireEvent({ eventName: TREE_EVENTS.onBlur, node: this });
        }
    };
    // 双击事件
    // 双击事件
    /**
     * @param {?} rawEvent
     * @return {?}
     */
    TreeNode.prototype.doublClick = 
    // 双击事件
    /**
     * @param {?} rawEvent
     * @return {?}
     */
    function (rawEvent) {
        this.fireEvent({ eventName: TREE_EVENTS.onDoubleClick, node: this, rawEvent: rawEvent });
    };
    // 右键快捷菜单
    // 右键快捷菜单
    /**
     * @param {?} rawEvent
     * @return {?}
     */
    TreeNode.prototype.contextMenu = 
    // 右键快捷菜单
    /**
     * @param {?} rawEvent
     * @return {?}
     */
    function (rawEvent) {
        if (this.enableCustomContextMenu()) { // 启用右键菜单功能后，禁用默认的右键菜单
            rawEvent.preventDefault();
        }
        this.fireEvent({ eventName: TREE_EVENTS.onContextMenu, node: this, rawEvent: rawEvent });
    };
    /**
     * @param {?} $event
     * @param {?=} data
     * @return {?}
     */
    TreeNode.prototype.dropMouseAction = /**
     * @param {?} $event
     * @param {?=} data
     * @return {?}
     */
    function ($event, data) {
        if (data === void 0) { data = null; }
        // 1.设置focus
        this.treeModel.setFocus(true);
        // 2.执行事件处理器
        this._dropEventHandler(this.treeModel, this, $event, data);
        // 3.完成drop操作后，重置drag状态
        this.treeModel.cancelDrag();
    };
    /**
     * @private
     * @param {?} tree
     * @param {?} node
     * @param {?} $event
     * @param {?} to
     * @return {?}
     */
    TreeNode.prototype._dropEventHandler = /**
     * @private
     * @param {?} tree
     * @param {?} node
     * @param {?} $event
     * @param {?} to
     * @return {?}
     */
    function (tree, node, $event, to) {
        tree.moveNode({ from: tree.getDragNode(), to: to });
        // console.log("dropEventHandler:"+tree.roots);
    };
    // 点击右键菜单项
    // 点击右键菜单项
    /**
     * @param {?} event
     * @return {?}
     */
    TreeNode.prototype.rightMenuClicked = 
    // 点击右键菜单项
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.fireEvent({ eventName: TREE_EVENTS.onRightMenuClicked, node: this, rawEvent: event });
    };
    return TreeNode;
}());
export { TreeNode };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TreeNode.prototype._isExpanded;
    /** @type {?} */
    TreeNode.prototype.treeModel;
    /** @type {?} */
    TreeNode.prototype.parent;
    /** @type {?} */
    TreeNode.prototype.level;
    /** @type {?} */
    TreeNode.prototype.isVirtualRoot;
    /** @type {?} */
    TreeNode.prototype.elementRef;
    /** @type {?} */
    TreeNode.prototype.parentNode;
    /** @type {?} */
    TreeNode.prototype._isActive;
    /** @type {?} */
    TreeNode.prototype.changeDisplayField;
    /** @type {?} */
    TreeNode.prototype.data;
}
