// @farris/ide-devkit/esm5/lib/components/ng2tree-common/models/tree-model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { TreeNode } from './tree-node';
import { TreeOptions } from './tree-options';
import { TREE_EVENTS } from '../constants/events';
import { first, last } from 'lodash-es';
var TreeModel = /** @class */ (function () {
    function TreeModel() {
        this.options = new TreeOptions();
        // focused Node may be not actived 
        // actived Node must be focused
        this.focusedNode = null; // be chosen node
        // be chosen node
        this.activeNode = null; // be chosen and actived node
        this.eventNames = Object.keys(TREE_EVENTS);
        this.firstUpdate = true;
        this._dragNode = null;
        this._dropLocation = null;
    }
    /**
     * @param {?} __0
     * @return {?}
     */
    TreeModel.prototype.setData = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var nodes = _a.nodes, options = _a.options, events = _a.events;
        this.options = new TreeOptions(options);
        this.events = events;
        this.update(nodes);
    };
    /**
     * @param {?} nodes
     * @return {?}
     */
    TreeModel.prototype.update = /**
     * @param {?} nodes
     * @return {?}
     */
    function (nodes) {
        // Update the tree:
        var _this = this;
        this.virtualRoot = new TreeNode({ isVirtualRoot: true }, null, this);
        this.roots = nodes && nodes.map(function (child) { return new TreeNode(child, _this.virtualRoot, _this); });
        this.virtualRoot[this.options.childrenField] = this.roots;
        this._loadTreeNodeContentComponent();
        // Fire event:
        if (this.firstUpdate) {
            if (this.roots) {
                this.fireEvent({ eventName: TREE_EVENTS.onInitialized });
                this.firstUpdate = false;
            }
        }
        else {
            this.fireEvent({ eventName: TREE_EVENTS.onUpdateData });
        }
    };
    //Used for code test
    //Used for code test
    /**
     * @return {?}
     */
    TreeModel.prototype.addStaticTreeNode = 
    //Used for code test
    /**
     * @return {?}
     */
    function () {
        this.createAndAddTreeNode({
            id: 1,
            name: 'root1',
            subTitle: 'the root',
            type: 'type1'
        }, this.focusedNode);
    };
    /**
     * @param {?} data
     * @param {?} parentNode
     * @param {?=} index
     * @return {?}
     */
    TreeModel.prototype.createAndAddTreeNode = /**
     * @param {?} data
     * @param {?} parentNode
     * @param {?=} index
     * @return {?}
     */
    function (data, parentNode, index) {
        /** @type {?} */
        var createdNode = this.createTreeNode(data, parentNode);
        this.addTreeNode(createdNode, parentNode, index);
    };
    /**
     * @param {?} data
     * @param {?} parent
     * @return {?}
     */
    TreeModel.prototype.createTreeNode = /**
     * @param {?} data
     * @param {?} parent
     * @return {?}
     */
    function (data, parent) {
        /** @type {?} */
        var createdNode = new TreeNode(data, parent, this);
        return createdNode;
    };
    /**
     * @param {?} addedNode
     * @param {?} parentNode
     * @param {?=} index
     * @return {?}
     */
    TreeModel.prototype.addTreeNode = /**
     * @param {?} addedNode
     * @param {?} parentNode
     * @param {?=} index
     * @return {?}
     */
    function (addedNode, parentNode, index) {
        if (addedNode == null) {
            return;
        }
        if (parentNode == null) { // 增加顶级树节点（没有父节点的树节点）
            this.roots.push(addedNode);
        }
        else {
            if (index === null || index === undefined) {
                parentNode.childrenField.push(addedNode);
            }
            else {
                parentNode.childrenField.splice(index, 0, addedNode);
            }
        }
        this.update(this.roots);
        this.fireEvent({ eventName: TREE_EVENTS.onAddNode, addedNode: addedNode, parentNode: parentNode });
    };
    /**
     * @return {?}
     */
    TreeModel.prototype.removeFocusedTreeNode = /**
     * @return {?}
     */
    function () {
        this.removeTreeNode(this.focusedNode);
    };
    //移除选中的已知节点
    //移除选中的已知节点
    /**
     * @param {?} selectedTreeNode
     * @return {?}
     */
    TreeModel.prototype.removeTreeNode = 
    //移除选中的已知节点
    /**
     * @param {?} selectedTreeNode
     * @return {?}
     */
    function (selectedTreeNode) {
        if (selectedTreeNode == null) {
            return;
        }
        /** @type {?} */
        var parent = selectedTreeNode.parent;
        if (parent == null) { //移除顶级树节点（没有父节点的树节点）
            //移除顶级树节点（没有父节点的树节点）
            /** @type {?} */
            var index = this.roots.indexOf(selectedTreeNode);
            this.roots.splice(index, 1); //移除数组中某一指定节点
        }
        else {
            if (parent.childrenField.length <= 0) {
                console.log("RemoveTreeNode Warning: it is impossible to remove element from an empty array");
                return;
            }
            //移除数组中某一指定节点
            /** @type {?} */
            var index = parent.childrenField.indexOf(selectedTreeNode);
            parent.childrenField.splice(index, 1);
        }
        this.update(this.roots);
        this.fireEvent({ eventName: TREE_EVENTS.onRemoveNode, selectedTreeNode: selectedTreeNode, parent: parent });
    };
    /**
     * 定位指定树节点
     * @param needLocatedNode 待定位节点
     */
    /**
     * 定位指定树节点
     * @param {?} needLocatedNode 待定位节点
     * @return {?}
     */
    TreeModel.prototype.locateTreeNode = /**
     * 定位指定树节点
     * @param {?} needLocatedNode 待定位节点
     * @return {?}
     */
    function (needLocatedNode) {
        if (needLocatedNode == null) {
            return;
        }
        /** @type {?} */
        var parentNode = needLocatedNode.parent;
        while (parentNode != null) {
            parentNode.isExpanded = true;
            parentNode = parentNode.parent;
        }
        needLocatedNode.isActive = false;
        needLocatedNode.toggleActivated();
    };
    /**
     * 通过ID定位树节点
     * @param nodeID 待查找并定位节点的ID
     */
    /**
     * 通过ID定位树节点
     * @param {?} nodeID 待查找并定位节点的ID
     * @return {?}
     */
    TreeModel.prototype.locateNodeByID = /**
     * 通过ID定位树节点
     * @param {?} nodeID 待查找并定位节点的ID
     * @return {?}
     */
    function (nodeID) {
        /** @type {?} */
        var node = this.searchTreeNodeByID(nodeID);
        if (node == null) {
            return false;
        }
        this.locateTreeNode(node);
        return true;
    };
    /**
     * 查找指定树节点
     * @param nodeID 待查找节点ID
     */
    /**
     * 查找指定树节点
     * @param {?} nodeID 待查找节点ID
     * @return {?}
     */
    TreeModel.prototype.searchTreeNodeByID = /**
     * 查找指定树节点
     * @param {?} nodeID 待查找节点ID
     * @return {?}
     */
    function (nodeID) {
        return this.searchTreeNode(this.roots, nodeID);
    };
    /**
     * 在指定集合中，根据ID查找树节点
     * @param nodes 树集合
     * @param nodeID 待查找节点ID
     */
    /**
     * 在指定集合中，根据ID查找树节点
     * @param {?} nodes 树集合
     * @param {?} nodeID 待查找节点ID
     * @return {?}
     */
    TreeModel.prototype.searchTreeNode = /**
     * 在指定集合中，根据ID查找树节点
     * @param {?} nodes 树集合
     * @param {?} nodeID 待查找节点ID
     * @return {?}
     */
    function (nodes, nodeID) {
        var _this = this;
        if (nodes == null || nodes.length <= 0) {
            return null;
        }
        if (nodeID == null || nodeID.length < 0) {
            return null;
        }
        /** @type {?} */
        var searchedTreeNode = null;
        nodes.forEach(function (node) {
            if (node.idField == nodeID) { //回归
                searchedTreeNode = node;
                return;
            }
            if (node.childrenField == null || node.childrenField.length < 0) { //回归
                return;
            }
            /** @type {?} */
            var searchedNodeInChildren = _this.searchTreeNode(node.childrenField, nodeID);
            if (searchedNodeInChildren != null) {
                searchedTreeNode = searchedNodeInChildren;
            }
            return; //回归
        });
        return searchedTreeNode; //返回
    };
    Object.defineProperty(TreeModel.prototype, "treeNodeContentComponent", {
        get: /**
         * @return {?}
         */
        function () { return this._treeNodeContentComponent; },
        enumerable: true,
        configurable: true
    });
    ;
    // if treeNodeTemplate is a component - use it,
    // otherwise - it's a template, so wrap it with an AdHoc component
    // if treeNodeTemplate is a component - use it,
    // otherwise - it's a template, so wrap it with an AdHoc component
    /**
     * @return {?}
     */
    TreeModel.prototype._loadTreeNodeContentComponent = 
    // if treeNodeTemplate is a component - use it,
    // otherwise - it's a template, so wrap it with an AdHoc component
    /**
     * @return {?}
     */
    function () {
        this._treeNodeContentComponent = this.options.treeNodeTemplate;
        if (typeof this._treeNodeContentComponent === 'string') {
            // this._treeNodeContentComponent = this._createAdHocComponent(this._treeNodeContentComponent);
        }
    };
    Object.defineProperty(TreeModel.prototype, "isFocused", {
        // _createAdHocComponent(templateStr) {
        //     @Component({
        //         selector: 'TreeNodeTemplate',
        //         template: templateStr
        //     })
        //     class AdHocTreeNodeTemplateComponent {
        //         @Input() node: TreeNode;
        //     }
        //     return AdHocTreeNodeTemplateComponent;
        // }
        get: 
        // _createAdHocComponent(templateStr) {
        //     @Component({
        //         selector: 'TreeNodeTemplate',
        //         template: templateStr
        //     })
        //     class AdHocTreeNodeTemplateComponent {
        //         @Input() node: TreeNode;
        //     }
        //     return AdHocTreeNodeTemplateComponent;
        // }
        /**
         * @return {?}
         */
        function () {
            return TreeModel.focusedTree === this;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} value
     * @return {?}
     */
    TreeModel.prototype.setFocus = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        TreeModel.focusedTree = value ? this : null;
    };
    /**
     * @return {?}
     */
    TreeModel.prototype.getFirstRoot = /**
     * @return {?}
     */
    function () {
        return first(this.roots);
    };
    /**
     * @return {?}
     */
    TreeModel.prototype.getLastRoot = /**
     * @return {?}
     */
    function () {
        return last(this.roots);
    };
    /**
     * @return {?}
     */
    TreeModel.prototype.focusNextNode = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var previousNode = this.focusedNode;
        /** @type {?} */
        var nextNode = previousNode ? previousNode.findNextNode() : this.getFirstRoot();
        nextNode && nextNode.focus(); // Short-circuit evaluation
    };
    /**
     * @return {?}
     */
    TreeModel.prototype.focusPreviousNode = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var previousNode = this.focusedNode;
        /** @type {?} */
        var nextNode = previousNode ? previousNode.findPreviousNode() : this.getLastRoot();
        nextNode && nextNode.focus();
    };
    /**
     * @return {?}
     */
    TreeModel.prototype.focusDrillUp = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var previousNode = this.focusedNode;
        /** @type {?} */
        var nextNode = previousNode && previousNode.realParent;
        nextNode && nextNode.focus();
    };
    /**
     * @return {?}
     */
    TreeModel.prototype.focusDrillDown = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var previousNode = this.focusedNode;
        /** @type {?} */
        var nextNode = previousNode && previousNode.getFirstChild();
        nextNode && nextNode.focus();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TreeModel.prototype.fireEvent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // https://stackoverflow.com/questions/35840576/differencse-between-eventemitter-next-and-eventemitter-emit-in-angular-2
        //  abandon next() function, begin to use emit() function
        // this.events[event.eventName].next(event);
        // this.events[event.eventName].emit(event,alert(event.eventName));//发射事件，并传递事件的对象
        this.events[event.eventName].emit(event);
    };
    /**
     * 判断是否执行移动节点操作，可以移动返回true，否则返回false
     * @param param0 param0.from 待移动节点的原有父节点
     *               param0.to 待移动节点的新父节点
     */
    /**
     * 判断是否执行移动节点操作，可以移动返回true，否则返回false
     * @param {?} __0
     * @return {?}
     */
    TreeModel.prototype.canMoveNode = /**
     * 判断是否执行移动节点操作，可以移动返回true，否则返回false
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var from = _a.from, to = _a.to;
        // same node
        if (from.parentNode === to.parentNode && from.index === to.index) {
            return false;
        }
        /** @type {?} */
        var fromChildren = from.parentNode.children;
        /** @type {?} */
        var fromNode = fromChildren[from.index];
        return !to.parentNode.isDescendantOf(fromNode);
    };
    /**
     * 移动节点
     * @param param0 param0.from 待移动节点的原有父节点
     *               param0.to 待移动节点的新父节点
     */
    /**
     * 移动节点
     * @param {?} __0
     * @return {?}
     */
    TreeModel.prototype.moveNode = /**
     * 移动节点
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var from = _a.from, to = _a.to;
        if (!this.canMoveNode({ from: from, to: to }))
            return;
        /** @type {?} */
        var fromChildren = from.parentNode.childrenField;
        // If node doesn't have children - create children array
        if (!to.parentNode.childrenField) {
            to.parentNode.childrenField = [];
        }
        /** @type {?} */
        var toChildren = to.parentNode.childrenField;
        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice
        // The splice() method changes the contents of an array by removing existing elements and/or adding new elements.
        /** @type {?} */
        var node = fromChildren.splice(from.index, 1)[0];
        // Compensate for index if already removed from parent:
        /** @type {?} */
        var toIndex = (from.parentNode === to.parentNode && to.index > from.index) ? to.index - 1 : to.index;
        toChildren.splice(toIndex, 0, node);
        // console.log("toChildren:" + toChildren);
        // console.log("AfterMoveNode:" + this.roots);
        this.update(this.roots); // 实现node moved后，重新刷新这棵树
        this.fireEvent({ eventName: TREE_EVENTS.onMoveNode, node: node, to: to });
    };
    // TODO: move to a different service:
    // TODO: move to a different service:
    /**
     * @param {?} dragNode
     * @return {?}
     */
    TreeModel.prototype.setDragNode = 
    // TODO: move to a different service:
    /**
     * @param {?} dragNode
     * @return {?}
     */
    function (dragNode) {
        this._dragNode = dragNode;
    };
    /**
     * @return {?}
     */
    TreeModel.prototype.getDragNode = /**
     * @return {?}
     */
    function () {
        return this._dragNode || { parentNode: null, index: null };
    };
    /**
     * @return {?}
     */
    TreeModel.prototype.isDragging = /**
     * @return {?}
     */
    function () {
        return this.getDragNode().parentNode;
    };
    /**
     * @param {?} dropLocation
     * @return {?}
     */
    TreeModel.prototype.setDropLocation = /**
     * @param {?} dropLocation
     * @return {?}
     */
    function (dropLocation) {
        this._dropLocation = dropLocation;
    };
    /**
     * @return {?}
     */
    TreeModel.prototype.getDropLocation = /**
     * @return {?}
     */
    function () {
        return this._dropLocation || { component: null, parentNode: null, index: null };
    };
    /**
     * @param {?} component
     * @return {?}
     */
    TreeModel.prototype.isDraggingOver = /**
     * @param {?} component
     * @return {?}
     */
    function (component) {
        return this.getDropLocation().component === component;
    };
    /**
     * @return {?}
     */
    TreeModel.prototype.cancelDrag = /**
     * @return {?}
     */
    function () {
        this.setDropLocation(null);
        this.setDragNode(null);
    };
    TreeModel.focusedTree = null;
    TreeModel.decorators = [
        { type: Injectable }
    ];
    return TreeModel;
}());
export { TreeModel };
if (false) {
    /** @type {?} */
    TreeModel.focusedTree;
    /** @type {?} */
    TreeModel.prototype.roots;
    /** @type {?} */
    TreeModel.prototype.options;
    /** @type {?} */
    TreeModel.prototype.focusedNode;
    /** @type {?} */
    TreeModel.prototype.activeNode;
    /**
     * @type {?}
     * @private
     */
    TreeModel.prototype.events;
    /** @type {?} */
    TreeModel.prototype.eventNames;
    /** @type {?} */
    TreeModel.prototype.firstUpdate;
    /** @type {?} */
    TreeModel.prototype._dragNode;
    /** @type {?} */
    TreeModel.prototype._dropLocation;
    /** @type {?} */
    TreeModel.prototype.virtualRoot;
    /**
     * @type {?}
     * @private
     */
    TreeModel.prototype._treeNodeContentComponent;
    /* Skipping unhandled member: ;*/
}
