// @gspwidget/widget-core/esm2015/lib/widget/runtime-agg.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PlThemeColors } from "@gspwidget/util";
/** @type {?} */
const aggFirstDim2Field = (/**
 * @param {?} runtimeAgg
 * @param {?} allDims
 * @return {?}
 */
(runtimeAgg, allDims) => getFieldByName(allDims, runtimeAgg.dimensions[0]));
const ɵ0 = aggFirstDim2Field;
/** @type {?} */
const aggFirstInd2Field = (/**
 * @param {?} runtimeAgg
 * @param {?} allDims
 * @return {?}
 */
(runtimeAgg, allDims) => getFieldByName(allDims, runtimeAgg.indicators[0].dimName));
const ɵ1 = aggFirstInd2Field;
/** @type {?} */
const stackAggFields = (/**
 * @param {?=} pure
 * @return {?}
 */
(pure = false) => [{
        name: 'yFieldObjs',
        valueFactory: (/**
         * @param {?} __0
         * @param {?} allDims
         * @return {?}
         */
        ({ indicators }, allDims) => {
            return indicators.map((/**
             * @param {?} ind
             * @param {?} i
             * @return {?}
             */
            (ind, i) => ({
                valueField: getFieldByName(allDims, ind.dimName),
                color: pure
                    ? PlThemeColors.pure[i % indicators.length]
                    : PlThemeColors.gradient[i % indicators.length]
            })));
        })
    }, {
        name: 'categoryFieldName',
        valueFactory: aggFirstDim2Field
    }]);
const ɵ2 = stackAggFields;
/** @type {?} */
const barAggFields = [
    { name: 'categoryFieldName', valueFactory: aggFirstDim2Field },
    { name: 'valueFieldName', valueFactory: aggFirstInd2Field }
];
/** @type {?} */
export const chartAggFieldMap = {
    stack: stackAggFields(),
    area: stackAggFields(true),
    'bar-group': stackAggFields(),
    'widget-line': stackAggFields(true),
    bar: barAggFields,
    pie: barAggFields,
    pie2: barAggFields,
    'bar-horizontal': barAggFields,
    'simple-line': barAggFields,
    'bar-multiple': [{
            categoryFieldName: aggFirstDim2Field,
            valueFieldNameOne: aggFirstInd2Field,
            valueFieldNameTwo: (/**
             * @param {?} runtimeAgg
             * @param {?} allDims
             * @return {?}
             */
            (runtimeAgg, allDims) => getFieldByName(allDims, runtimeAgg.indicators[1].dimName)),
        }]
};
/**
 * @param {?} dims
 * @param {?} dimName
 * @return {?}
 */
function getFieldByName(dims, dimName) {
    const { name, label } = dims.find((/**
     * @param {?} dim
     * @return {?}
     */
    dim => dim.name === dimName));
    return { name, label };
}
/**
 * @param {?} widgetName
 * @return {?}
 */
export function isWidgetRuntimeAggActive(widgetName) {
    return Object
        .keys(chartAggFieldMap)
        .includes(widgetName);
}
export { ɵ0, ɵ1, ɵ2 };
