// @farris/ui-notify/esm5/lib/notifiy.options.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
var NotifyOptions = /** @class */ (function () {
    function NotifyOptions() {
    }
    NotifyOptions.decorators = [
        { type: Injectable }
    ];
    return NotifyOptions;
}());
export { NotifyOptions };
if (false) {
    /** @type {?} */
    NotifyOptions.prototype.type;
    /** @type {?} */
    NotifyOptions.prototype.title;
    /** @type {?} */
    NotifyOptions.prototype.msg;
    /** @type {?} */
    NotifyOptions.prototype.showClose;
    /** @type {?} */
    NotifyOptions.prototype.theme;
    /** @type {?} */
    NotifyOptions.prototype.timeout;
    /** @type {?} */
    NotifyOptions.prototype.onAdd;
    /** @type {?} */
    NotifyOptions.prototype.onRemove;
}
var NotifyData = /** @class */ (function (_super) {
    tslib_1.__extends(NotifyData, _super);
    function NotifyData() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return NotifyData;
}(NotifyOptions));
export { NotifyData };
if (false) {
    /** @type {?} */
    NotifyData.prototype.id;
}
var NotifyConfig = /** @class */ (function () {
    function NotifyConfig() {
        // Maximum number of toasties to show at once
        this.limit = 5;
        // Whether to show the 'X' icon to close the toast
        this.showClose = true;
        // The window position where the toast pops up
        this.position = 'top-center';
        // How long (in miliseconds) the toasty shows before it's removed. Set to null/0 to turn off.
        this.timeout = 3000;
        // What theme to use
        this.theme = 'bootstrap';
    }
    NotifyConfig.decorators = [
        { type: Injectable }
    ];
    return NotifyConfig;
}());
export { NotifyConfig };
if (false) {
    /** @type {?} */
    NotifyConfig.prototype.limit;
    /** @type {?} */
    NotifyConfig.prototype.showClose;
    /** @type {?} */
    NotifyConfig.prototype.position;
    /** @type {?} */
    NotifyConfig.prototype.timeout;
    /** @type {?} */
    NotifyConfig.prototype.theme;
    /** @type {?} */
    NotifyConfig.prototype.left;
    /** @type {?} */
    NotifyConfig.prototype.right;
    /** @type {?} */
    NotifyConfig.prototype.top;
    /** @type {?} */
    NotifyConfig.prototype.bottom;
}
/** @enum {number} */
var NotifyEventType = {
    ADD: 0,
    CLEAR: 1,
    CLEAR_ALL: 2,
};
export { NotifyEventType };
NotifyEventType[NotifyEventType.ADD] = 'ADD';
NotifyEventType[NotifyEventType.CLEAR] = 'CLEAR';
NotifyEventType[NotifyEventType.CLEAR_ALL] = 'CLEAR_ALL';
var NotifyEvent = /** @class */ (function () {
    function NotifyEvent(type, value) {
        this.type = type;
        this.value = value;
    }
    return NotifyEvent;
}());
export { NotifyEvent };
if (false) {
    /** @type {?} */
    NotifyEvent.prototype.type;
    /** @type {?} */
    NotifyEvent.prototype.value;
}
