// @gsp-sys/rac-common/esm5/lib/component/reassign/reassign-grid/reassign-grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, TemplateRef } from '@angular/core';
import { ReAssignService } from '../../../common/service/reassign.service';
var ReassignGridComponent = /** @class */ (function () {
    function ReassignGridComponent(reAssignService) {
        this.reAssignService = reAssignService;
        this.selectedKey = '';
        this.selectedKeyOld = '';
        this.selectedChanged = new EventEmitter();
        /**
         * 帮助输出事件
         */
        this.selectedAfterOk = new EventEmitter();
        this.closeModalEmitter = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ReassignGridComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.columns = [
            // { field: 'name', title: this.translate.instant('lblAuthObj') }
            { field: 'authorizationCode', title: '权限对象编号', width: '0' },
            { field: 'authorizationName', title: '权限对象名称', width: '200' },
            { field: 'authfieldCode', title: '权限字段编号', width: '0' },
            { field: 'authfieldName', title: '权限字段名称' }
        ];
    };
    /** 获取允许重授权的权限 */
    /**
     * 获取允许重授权的权限
     * @return {?}
     */
    ReassignGridComponent.prototype.initData = /**
     * 获取允许重授权的权限
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var obs = this.reAssignService.getAll();
        obs.subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            _this.datas = result;
        }));
    };
    /** 输入属性变化后 */
    /**
     * 输入属性变化后
     * @param {?} changes
     * @return {?}
     */
    ReassignGridComponent.prototype.ngOnChanges = /**
     * 输入属性变化后
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (!this.datas || this.datas.length <= 0) {
            // 服务端获取数据
            this.initData();
        }
    };
    /**
     * 选中行改变后触发事件
     */
    /**
     * 选中行改变后触发事件
     * @param {?} args
     * @return {?}
     */
    ReassignGridComponent.prototype.handleSelection = /**
     * 选中行改变后触发事件
     * @param {?} args
     * @return {?}
     */
    function (args) {
        this.selectedModel = args.data;
        if (!this.selectedModel) {
            this.selectedKey = '';
            return;
        }
        this.selectedKey = this.selectedModel.id;
        // 首次触发
        if (this.selectedKeyOld === '') {
            this.selectedKeyOld = this.selectedKey;
            this.selectedChanged.emit(this.selectedModel);
        }
        else {
            // 变化的情况下触发
            if (this.selectedKeyOld !== this.selectedKey) {
                this.selectedKeyOld = this.selectedKey;
                this.selectedChanged.emit(this.selectedModel);
            }
        }
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    ReassignGridComponent.prototype.iconClass = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var id = _a.id;
        return {
            'k-i-folder': true,
            'k-icon': true
        };
    };
    /** 帮助确定按钮 */
    /**
     * 帮助确定按钮
     * @return {?}
     */
    ReassignGridComponent.prototype.confirm = /**
     * 帮助确定按钮
     * @return {?}
     */
    function () {
        this.selectedKeyOld = this.selectedKey;
        this.selectedAfterOk.emit(this.selectedModel);
    };
    /**
     * @return {?}
     */
    ReassignGridComponent.prototype.cancle = /**
     * @return {?}
     */
    function () {
        this.closeModalEmitter.next();
    };
    ReassignGridComponent.decorators = [
        { type: Component, args: [{
                    selector: 'rac-reassign-grid',
                    template: "\r\n<div class=\"f-utils-flex-column\" style=\"height:450px;\">\r\n    <farris-datagrid   #farrisGrid  class=\"f-utils-flex-column f-utils-fill\" [columns]=\"columns\" [data]=\"datas\" [showBorder]=\"false\"\r\n        [idField]=\"'id'\"\r\n        [showLineNumber]=\"true\" [fitColumns]=\"true\" [fit]=\"true\" [striped]=\"true\" \r\n        (selectChanged)=\"handleSelection($event)\">\r\n    </farris-datagrid>\r\n</div>\r\n\r\n<ng-template #btnOkAndCancle >\r\n      <button class=\"btn btn-sm btn-primary\" (click)=\"confirm()\">\u786E\u8BA4</button>\r\n      <button class=\"btn btn-sm btn-default\" (click)=\"cancle()\">\u53D6\u6D88</button>\r\n</ng-template>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ReassignGridComponent.ctorParameters = function () { return [
        { type: ReAssignService }
    ]; };
    ReassignGridComponent.propDecorators = {
        datas: [{ type: Input }],
        selectedChanged: [{ type: Output }],
        selectedAfterOk: [{ type: Output }],
        closeModalEmitter: [{ type: Output }],
        btnOkAndCancleRef: [{ type: ViewChild, args: ['btnOkAndCancle',] }]
    };
    return ReassignGridComponent;
}());
export { ReassignGridComponent };
if (false) {
    /** @type {?} */
    ReassignGridComponent.prototype.columns;
    /** @type {?} */
    ReassignGridComponent.prototype.datas;
    /** @type {?} */
    ReassignGridComponent.prototype.selectedKey;
    /** @type {?} */
    ReassignGridComponent.prototype.selectedKeyOld;
    /** @type {?} */
    ReassignGridComponent.prototype.selectedModel;
    /** @type {?} */
    ReassignGridComponent.prototype.selectedChanged;
    /**
     * 帮助输出事件
     * @type {?}
     */
    ReassignGridComponent.prototype.selectedAfterOk;
    /** @type {?} */
    ReassignGridComponent.prototype.closeModalEmitter;
    /** @type {?} */
    ReassignGridComponent.prototype.btnOkAndCancleRef;
    /**
     * @type {?}
     * @private
     */
    ReassignGridComponent.prototype.reAssignService;
}
