// @farris/ide-devkit/esm2015/lib/components/angular-editor/angular-monaco-editor/angular-monaco-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { forwardRef, Inject, NgZone } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { fromEvent } from 'rxjs/observable/fromEvent';
import { ANGULAR_MONACO_EDITOR_CONFIG } from '../model/config';
import { CodeEditorEventService } from '../service/code-editor.event.service';
import { CODE_EDITOR_EVENTS } from '../constant/events';
import { BaseMonacoEditor } from '../model/base-monaco-editor';
import { AngularMonacoEditorService } from '../service/angular-monaco-editor.service';
// 自定义输入控件:1.封装ControlValueAccessor
// https://code-examples.net/zh-CN/q/2154761
/** @type {?} */
export const CODE_EDITOR_INPUT_VALUE_ACCESSOR = {
    // https://blog.csdn.net/wangdan_2013/article/details/81314959
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => AngularMonacoEditorComponent),
    multi: true
};
// 自定义输入控件 <-> Monaco Edtor
// 自定义输入控件:3.1 implements ControlValueAccessor
export class AngularMonacoEditorComponent extends BaseMonacoEditor {
    // tslint:disable-next-line:max-line-length
    /**
     * @param {?} angularMonacoEditorService
     * @param {?} zone
     * @param {?} editorEventService
     * @param {?} angularEditorconfig
     */
    constructor(angularMonacoEditorService, zone, editorEventService, angularEditorconfig) {
        super(editorEventService, angularEditorconfig);
        this.angularMonacoEditorService = angularMonacoEditorService;
        this.zone = zone;
        this.editorEventService = editorEventService;
        this.angularEditorconfig = angularEditorconfig;
        this._value = '';
        // ControlValueAccessor提供的事件回调
        this.onControlValueChange = (_) => {
        };
        // ControlValueAccessor提供的事件回调
        this.onControlTouched = () => {
        };
    }
    /**
     * @param {?} model
     * @return {?}
     */
    set model(model) {
        this.options.model = model;
        if (this._editor) {
            this._editor.dispose();
            this.initMonaco(this.options);
        }
    }
    // todo: 提取options公共类
    /**
     * @protected
     * @param {?} options
     * @return {?}
     */
    initMonaco(options) {
        /** @type {?} */
        const enableModel = !!options.model;
        if (enableModel) {
            /** @type {?} */
            const searchedModel = monaco.editor.getModel(options.model.uri);
            if (null !== searchedModel) {
                searchedModel.dispose();
            }
            options.model = monaco.editor.createModel(options.model.value, options.model.language, options.model.uri);
        }
        this._editor = monaco.editor.create(this._editorComponent.nativeElement, options);
        if (!enableModel) {
            this._editor.setValue(this._value);
        }
        if (enableModel) {
            this.angularMonacoEditorService.handleModelMarkers();
        }
        // monaco editor -> outside component
        this._editor.onDidChangeModelContent((e) => this.onChangeModelContentHandler(e));
        this._editor.onDidBlurEditorText((e) => this.onBlurEditorTextHandler(e));
        this._editor.onDidLayoutChange((e) => this.onLayoutChangeHandler(e));
        // refresh layout on resize event.
        this.refreshLayoutWhenWindowResize();
        this.editorEventService.fireEvent({
            eventName: CODE_EDITOR_EVENTS.onInit,
            target: this,
            editor: this._editor
        });
    }
    /**
     * refresh layout when resized the window
     * @return {?}
     */
    refreshLayoutWhenWindowResize() {
        if (this._windowResizeSubscription) {
            this._windowResizeSubscription.unsubscribe();
        }
        // fromEvent用于兼听事件，事件触发时，将事件event转成可流动的Observable进行传输
        // https://www.jianshu.com/p/46894deb870a
        this._windowResizeSubscription = fromEvent(window, 'resize').subscribe(() => this._editor.layout());
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onChangeModelContentHandler(e) {
        /** @type {?} */
        const _value = this._editor.getValue();
        // monaco editor -> outside component
        // https://github.com/JTangming/tm/issues/4 ngZone详解
        this.zone.run(() => this.value = _value); // value is not propagated to parent when executing outside zone.
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onBlurEditorTextHandler(e) {
        this.onControlTouched();
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onLayoutChangeHandler(e) {
        console.log('Layout changed:\n' + e);
    }
    // get accessor
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    // set accessor including call the onchange callback
    /**
     * @param {?} v
     * @return {?}
     */
    set value(v) {
        if (v !== this.value) { // 注意这种写法，值得学习
            this._value = v;
        }
        this.onControlValueChange(this.value); // 在属性修饰器里调用onControlValueChange方法
    }
    // 自定义输入控件:3.2 implements ControlValueAccesso
    // outside component -> monaco editor
    // From ControlValueAccessor interface
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        /** @type {?} */
        const self = this;
        self.value = value || '';
        // Fix for value change while dispose in process.
        setTimeout(() => {
            if (self._editor /*&& !this.options.model*/) {
                self._editor.setValue(self._value);
            }
        });
    }
    // From ControlValueAccessor interface
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onControlValueChange = fn;
    }
    // From ControlValueAccessor interface
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onControlTouched = fn;
    }
}
AngularMonacoEditorComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'angular-monaco-editor',
                template: "<div class=\"editor\" #codeEditor></div>",
                // 自定义输入控件:2.引入依赖服务ControlValueAccessor
                providers: [
                    CODE_EDITOR_INPUT_VALUE_ACCESSOR,
                    CodeEditorEventService,
                    AngularMonacoEditorService
                ],
                styles: [":host{display:block;min-height:200px}.editor{width:100%;height:100%}"]
            }] }
];
/** @nocollapse */
AngularMonacoEditorComponent.ctorParameters = () => [
    { type: AngularMonacoEditorService },
    { type: NgZone },
    { type: CodeEditorEventService },
    { type: undefined, decorators: [{ type: Inject, args: [ANGULAR_MONACO_EDITOR_CONFIG,] }] }
];
AngularMonacoEditorComponent.propDecorators = {
    model: [{ type: Input, args: ['model',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    AngularMonacoEditorComponent.prototype._value;
    /** @type {?} */
    AngularMonacoEditorComponent.prototype.onControlValueChange;
    /** @type {?} */
    AngularMonacoEditorComponent.prototype.onControlTouched;
    /**
     * @type {?}
     * @private
     */
    AngularMonacoEditorComponent.prototype.angularMonacoEditorService;
    /**
     * @type {?}
     * @private
     */
    AngularMonacoEditorComponent.prototype.zone;
    /**
     * @type {?}
     * @private
     */
    AngularMonacoEditorComponent.prototype.editorEventService;
    /**
     * @type {?}
     * @private
     */
    AngularMonacoEditorComponent.prototype.angularEditorconfig;
}
