// @farris/ui-filter-editor/esm2015/lib/components/filter-footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output, HostBinding } from '@angular/core';
import { FilterGridComponent } from './filter-grid.component';
export class FilterFooterComponent {
    /**
     * @param {?} grid
     */
    constructor(grid) {
        this.grid = grid;
        this.cls = 'f-filter-footer';
        this.addFilter = new EventEmitter();
        this.clearFilter = new EventEmitter();
        this.moveTop = new EventEmitter();
        this.movePrev = new EventEmitter();
        this.moveNext = new EventEmitter();
        this.moveBottom = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    onAddFilter() {
        this.addFilter.emit();
    }
    /**
     * @return {?}
     */
    onClear() {
        this.clearFilter.emit();
    }
    /**
     * @return {?}
     */
    onMoveTop() {
        this.moveTop.emit();
    }
    /**
     * @return {?}
     */
    onMovePrev() {
        this.movePrev.emit();
    }
    /**
     * @return {?}
     */
    onMoveNext() {
        this.moveNext.emit();
    }
    /**
     * @return {?}
     */
    onMoveBottom() {
        this.moveBottom.emit();
    }
    /**
     * @return {?}
     */
    canUse() {
        if (this.grid.currentCondition) {
            return this.grid.conditions.length > 1;
        }
        return false;
    }
    /**
     * @return {?}
     */
    canMoveUp() {
        if (this.canUse()) {
            return this.grid.currentCondition.index > 0;
        }
        return false;
    }
    /**
     * @return {?}
     */
    canMoveDown() {
        if (this.canUse()) {
            return this.grid.currentCondition.index < this.grid.conditions.length - 1;
        }
        return false;
    }
}
FilterFooterComponent.decorators = [
    { type: Component, args: [{
                selector: 'filter-footer',
                template: `
    <div class="footer-container px-2" style="height: 36px;">
        <button type="button" (click)="onAddFilter()"
            class="k-button k-button-icontext k-flat "><span class="f-icon f-icon-add"></span> {{ 'filterEditor.addWhere' | locale }}</button>
        <button type="button" (click)="onClear()" [disabled]="!grid.conditions.length"
            class="k-button k-button-icontext k-flat "><span class="f-icon f-icon-close"></span> {{ 'filterEditor.clear' | locale }}</button>
        <button type="button" [disabled]="!canMoveUp()" (click)="onMoveTop()"
            class="k-button k-button-icontext k-flat "><span class="f-icon f-icon-arrow-end-up"></span>
            {{'filterEditor.moveTop' | locale }}</button>
        <button type="button" [disabled]="!canMoveUp()" (click)="onMovePrev()"
            class="k-button k-button-icontext k-flat "><span class="f-icon f-icon-arrow-60-up"></span>
            {{'filterEditor.moveUp' | locale }}</button>
        <button type="button" [disabled]="!canMoveDown()" (click)="onMoveNext()"
            class="k-button k-button-icontext k-flat "><span class="f-icon f-icon-arrow-60-down"></span>
            {{'filterEditor.moveDown' | locale }}</button>
        <button type="button" [disabled]="!canMoveDown()" (click)="onMoveBottom()"
            class="k-button k-button-icontext k-flat"><span class="f-icon f-icon-arrow-end-down"></span>
            {{'filterEditor.moveBottom' | locale }}</button>
    </div>
    `
            }] }
];
/** @nocollapse */
FilterFooterComponent.ctorParameters = () => [
    { type: FilterGridComponent }
];
FilterFooterComponent.propDecorators = {
    cls: [{ type: HostBinding, args: ['class',] }],
    addFilter: [{ type: Output }],
    clearFilter: [{ type: Output }],
    moveTop: [{ type: Output }],
    movePrev: [{ type: Output }],
    moveNext: [{ type: Output }],
    moveBottom: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FilterFooterComponent.prototype.cls;
    /** @type {?} */
    FilterFooterComponent.prototype.addFilter;
    /** @type {?} */
    FilterFooterComponent.prototype.clearFilter;
    /** @type {?} */
    FilterFooterComponent.prototype.moveTop;
    /** @type {?} */
    FilterFooterComponent.prototype.movePrev;
    /** @type {?} */
    FilterFooterComponent.prototype.moveNext;
    /** @type {?} */
    FilterFooterComponent.prototype.moveBottom;
    /** @type {?} */
    FilterFooterComponent.prototype.grid;
}
