// @farris/ui-filter-editor/esm2015/lib/components/filter-code.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ElementRef } from '@angular/core';
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @Company: Inspur
 * @Version: v0.0.1
 * @Date: 2019-04-09 11:03:49
 * @LastEditTime: 2019-08-15 17:22:59
 */
import { Component, Input, HostBinding } from '@angular/core';
export class FilterCodeComponent {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        this.cls = 'd-flex flex-column flex-fill';
        this.conditions = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.el.nativeElement.style.overflow = 'auto';
    }
}
FilterCodeComponent.decorators = [
    { type: Component, args: [{
                selector: 'filter-code',
                template: `
    <div style="overflow:auto" [style.height.px]="height - 38">
        <pre class="flex-fill">
            <code>{{ conditions | json }}</code>
        </pre>
    </div>
    `,
                styles: [`
        pre {
            color: #d8d9d9;
            background: #424242;
            padding-left: 20px;
            padding-top: 20px;
            margin-bottom: 0;
            font-size: 16px;
        }
        `]
            }] }
];
/** @nocollapse */
FilterCodeComponent.ctorParameters = () => [
    { type: ElementRef }
];
FilterCodeComponent.propDecorators = {
    cls: [{ type: HostBinding, args: ['class',] }],
    height: [{ type: Input }],
    conditions: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FilterCodeComponent.prototype.cls;
    /** @type {?} */
    FilterCodeComponent.prototype.height;
    /** @type {?} */
    FilterCodeComponent.prototype.conditions;
    /**
     * @type {?}
     * @private
     */
    FilterCodeComponent.prototype.el;
}
