// @edp-pmf/bpmn-model/dist/cjs/lib/model/bpmndi/BPMNShape.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BPMNShape = void 0;
var tslib_1 = require("tslib");
var BpmnModelHelper_1 = require("../BpmnModelHelper");
var ExtensionMethod_1 = require("../ExtensionMethod");
var PalUserTaskAttribute_1 = require("../mex/PalUserTaskAttribute");
var LabeledShape_1 = require("../omgdi/LabeledShape");
var ValidationError_1 = require("../ValidationError");
var ValidationLevel_1 = require("../ValidationLevel");
var BPMNLabel_1 = require("./BPMNLabel");
var BPMNShape = /** @class */ (function (_super) {
    tslib_1.__extends(BPMNShape, _super);
    function BPMNShape(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.bpmndi.BPMNShape";
        return _this;
    }
    BPMNShape.prototype.NewInstance = function () {
        return new BPMNShape(this.Model);
    };
    BPMNShape.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.BpmnElementId = null;
        this.BpmnLabel = null;
    };
    BPMNShape.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "BpmnElementId", this.BpmnElementId);
        if (this.BpmnLabel != null) {
            BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "BpmnLabel", this.BpmnLabel.ConvertToJsonObject());
        }
        return obj;
    };
    BPMNShape.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.BpmnElementId = obj.BpmnElementId;
        if (obj.hasOwnProperty("BpmnLabel")) {
            this.BpmnLabel = new BPMNLabel_1.BPMNLabel(this.Model);
            this.BpmnLabel.LoadFromJsonObject(obj.BpmnLabel);
        }
    };
    BPMNShape.prototype.LoadExtensionElements = function (extensionElementsObj) {
        _super.prototype.LoadExtensionElements.call(this, extensionElementsObj);
        var palUserTaskAttributeObj = ExtensionMethod_1.ExtensionMethod.GetAttributeValue(extensionElementsObj, "palUserTaskAttribute");
        if (palUserTaskAttributeObj != null) {
            this.palUserTaskAttribute = new PalUserTaskAttribute_1.PalUserTaskAttribute(this.Model);
            this.palUserTaskAttribute.LoadFromJsonObject(palUserTaskAttributeObj);
        }
    };
    BPMNShape.prototype.WriteExtensionElements = function (extensionElementsObj) {
        _super.prototype.WriteExtensionElements.call(this, extensionElementsObj);
        if (this.palUserTaskAttribute != null) {
            ExtensionMethod_1.ExtensionMethod.SetAttributeValue(extensionElementsObj, "palUserTaskAttribute", this.palUserTaskAttribute.ConvertToJsonObject());
        }
    };
    BPMNShape.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
        if (this.Model.GetHasIdElement(this.BpmnElementId) == null) {
            var error = new ValidationError_1.ValidationError();
            error.Element = this;
            error.ErrorLevel = ValidationLevel_1.ValidationLevel.$_Error;
            error.ErrorCode = "";
            error.ErrorMessage = "\u627E\u4E0D\u5230 id \u4E3A " + this.BpmnElementId + " \u7684\u5143\u7D20";
            errors.push(error);
        }
    };
    BPMNShape.MEXTAG = "shape.bpa";
    return BPMNShape;
}(LabeledShape_1.LabeledShape));
exports.BPMNShape = BPMNShape;
