// @edp-pmf/bpmn-model/dist/cjs/lib/model/bpmndi/BPMNEdge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BPMNEdge = void 0;
var tslib_1 = require("tslib");
var BpmnModelHelper_1 = require("../BpmnModelHelper");
var RectangularLine_1 = require("../mex/RectangularLine");
var LabeledEdge_1 = require("../omgdi/LabeledEdge");
var ValidationError_1 = require("../ValidationError");
var ValidationLevel_1 = require("../ValidationLevel");
var BPMNLabel_1 = require("./BPMNLabel");
var SequenceFlowLineType_1 = require("./SequenceFlowLineType");
var BPMNEdge = /** @class */ (function (_super) {
    tslib_1.__extends(BPMNEdge, _super);
    function BPMNEdge(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.bpmndi.BPMNEdge";
        _this.LineType = SequenceFlowLineType_1.SequenceFlowLineType.PolyLine;
        return _this;
    }
    BPMNEdge.prototype.NewInstance = function () {
        return new BPMNEdge(this.Model);
    };
    BPMNEdge.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.BpmnElementId = null;
        this.LineType = SequenceFlowLineType_1.SequenceFlowLineType.PolyLine;
        this.SourceShapeElementId = null;
        this.TargetShapeElementId = null;
        this.BpmnLabel = null;
        this.RectangularLine = null;
    };
    BPMNEdge.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "BpmnElementId", this.BpmnElementId);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "SourceShapeElementId", this.SourceShapeElementId);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "TargetShapeElementId", this.TargetShapeElementId);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "LineType", this.LineType);
        if (this.BpmnLabel != null) {
            BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "BpmnLabel", this.BpmnLabel.ConvertToJsonObject());
        }
        return obj;
    };
    BPMNEdge.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.BpmnElementId = obj.BpmnElementId;
        this.SourceShapeElementId = obj.SourceShapeElementId;
        this.TargetShapeElementId = obj.TargetShapeElementId;
        this.LineType = obj.LineType;
        if (obj.hasOwnProperty("BpmnLabel")) {
            this.BpmnLabel = new BPMNLabel_1.BPMNLabel(this.Model);
            this.BpmnLabel.LoadFromJsonObject(obj.BpmnLabel);
        }
    };
    BPMNEdge.prototype.LoadExtensionElements = function (extensionElementsObj) {
        _super.prototype.LoadExtensionElements.call(this, extensionElementsObj);
        if (extensionElementsObj.hasOwnProperty("RectangularLine")) {
            this.RectangularLine = new RectangularLine_1.RectangularLine(this.Model);
            this.RectangularLine.LoadFromJsonObject(extensionElementsObj.RectangularLine);
        }
    };
    BPMNEdge.prototype.WriteExtensionElements = function (extensionElementsObj) {
        _super.prototype.WriteExtensionElements.call(this, extensionElementsObj);
        switch (this.LineType) {
            case SequenceFlowLineType_1.SequenceFlowLineType.RectangularLineType:
                BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(extensionElementsObj, "RectangularLine", this.RectangularLine.ConvertToJsonObject());
                break;
            default:
                break;
        }
    };
    BPMNEdge.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
        if (this.Model.GetHasIdElement(this.BpmnElementId) == null) {
            var error = new ValidationError_1.ValidationError();
            error.Element = this;
            error.ErrorLevel = ValidationLevel_1.ValidationLevel.$_Error;
            error.ErrorCode = "";
            error.ErrorMessage = "\u627E\u4E0D\u5230 id \u4E3A " + this.BpmnElementId + " \u7684\u5143\u7D20";
            errors.push(error);
        }
    };
    return BPMNEdge;
}(LabeledEdge_1.LabeledEdge));
exports.BPMNEdge = BPMNEdge;
