// @gsp-sys/rtf-common/esm2015/lib/services/frmVariableService/framework-variable.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { WindowVariableService } from '@ecp-caf/caf-common';
export class FrameworkVariableService {
    constructor() {
        this.frameworkVariable = 'frameworkVariable';
        this.nameSpace = 'variables';
        this.frmVariableName = 'rtfFrmVariable';
        this.windowService = new WindowVariableService();
        this.frmVariableInitial();
    }
    /**
     * @return {?}
     */
    getFrameworkVariable() {
        return this.windowService.getTopWindow()[this.frameworkVariable];
    }
    /**
     * 在框架顶级window注册变量
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    frameworkVariableReg(key, value) {
        /** @type {?} */
        const frmVariable = this.getFrameworkVariable();
        if (frmVariable && frmVariable.name && frmVariable.name === this.frmVariableName) {
            frmVariable[this.nameSpace] = frmVariable[this.nameSpace] || {};
            frmVariable[this.nameSpace][key] = value;
            return;
        }
        this.frmVariableInitial();
        this.getFrameworkVariable()[this.nameSpace][key] = value;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getFrmVariableByKey(key) {
        /** @type {?} */
        const frmVariable = this.getFrameworkVariable();
        if (frmVariable && frmVariable[this.nameSpace] && frmVariable[this.nameSpace][key]) {
            return frmVariable[this.nameSpace][key];
        }
        return null;
    }
    /**
     * 在框架的顶级window上
     * 定义一个公共变量。
     * 用于存放菜单、应用缓存数据，eventbus都公共对象
     * 对适配层提供服务
     * @private
     * @return {?}
     */
    frmVariableInitial() {
        /** @type {?} */
        const topWin = this.windowService.getTopWindow();
        /** @type {?} */
        const name = 'name';
        topWin[this.frameworkVariable] = topWin[this.frameworkVariable] || {};
        topWin[this.frameworkVariable][name] = this.frmVariableName;
        topWin[this.frameworkVariable][this.nameSpace] = topWin[this.frameworkVariable][this.nameSpace] || {};
    }
}
FrameworkVariableService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FrameworkVariableService.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    FrameworkVariableService.prototype.frameworkVariable;
    /**
     * @type {?}
     * @private
     */
    FrameworkVariableService.prototype.nameSpace;
    /**
     * @type {?}
     * @private
     */
    FrameworkVariableService.prototype.frmVariableName;
    /**
     * @type {?}
     * @private
     */
    FrameworkVariableService.prototype.windowService;
}
