// @gsp-wf/wf-process-editor/esm2015/lib/views/note/note.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, TemplateRef, Optional, ChangeDetectorRef, LOCALE_ID, Inject } from '@angular/core';
import { Subject } from 'rxjs';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../i18n/pipe/wf-process-editor-locale.pipe';
export class NoteComponent {
    /**
     * @param {?} cdr
     * @param {?} localeId
     * @param {?} resources
     * @param {?} localePipe
     */
    constructor(cdr, localeId, resources, localePipe) {
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.submit = new Subject();
        this.cancel = new Subject();
        this.placeholder = this.localePipe.transform('component.note.placeholder');
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    clickOK() {
        this.submit.next(this.record);
    }
    /**
     * @return {?}
     */
    clickCancel() {
        this.cancel.next();
    }
}
NoteComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-note',
                template: "<div class=\"f-page\">\r\n  <div class=\"f-page-main f-utils-fill-flex-column h-100\">\r\n    <div class=\"f-utils-fill-flex-column farris-overflow-y-auto h-100\">\r\n      <textarea class=\"form-control content-textarea\" style=\"height: 100%;\" [(ngModel)]=\"record\" [placeholder]=\"placeholder\"></textarea>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #footer>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm\" (click)=\"clickCancel()\">{{'component.note.clickCancel' | wfLocale}}</button>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">{{'component.note.clickOK' | wfLocale}}</button>\r\n</ng-template>\r\n",
                providers: [
                    WfLocalePipe,
                ],
                styles: [""]
            }] }
];
/** @nocollapse */
NoteComponent.ctorParameters = () => [
    { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
    { type: WfLocalePipe, decorators: [{ type: Optional }] }
];
NoteComponent.propDecorators = {
    footer: [{ type: ViewChild, args: ['footer',] }]
};
if (false) {
    /** @type {?} */
    NoteComponent.prototype.footer;
    /** @type {?} */
    NoteComponent.prototype.record;
    /** @type {?} */
    NoteComponent.prototype.submit;
    /** @type {?} */
    NoteComponent.prototype.cancel;
    /** @type {?} */
    NoteComponent.prototype.placeholder;
    /**
     * @type {?}
     * @private
     */
    NoteComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    NoteComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    NoteComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    NoteComponent.prototype.localePipe;
}
