// @farris/mobile-cli/lib/utils/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.exit = exports.isExists = exports.isArray = exports.isObject = exports.parseOption = exports.CWD = exports.ENV = void 0;
const fs_1 = require("fs");
exports.ENV = process.env.NODE_ENV;
exports.CWD = process.cwd();
exports.parseOption = (option) => {
    if (typeof option === 'string') {
        return {
            type: option
        };
    }
    return option;
};
function isObject(obj) {
    return Object.prototype.toString.call(obj) === '[object Object]';
}
exports.isObject = isObject;
function isArray(arr) {
    return Array.isArray(arr);
}
exports.isArray = isArray;
function isExists(path, suffixs = ['js', 'json']) {
    let result = fs_1.existsSync(path);
    if (!result) {
        for (let index = 0; index < suffixs.length; index++) {
            const suffix = suffixs[index];
            result = fs_1.existsSync(`${path}.${suffix}`);
            if (result) {
                break;
            }
        }
    }
    return result;
}
exports.isExists = isExists;
function exit(code = 0) {
    process.exit(code);
}
exports.exit = exit;
