// @gsp-wf/rtdevkit/esm2015/lib/runtime/components/task-assign/task-assign.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { TaskService } from '../../services/task.service';
import { UserService } from '../../services/user.service';
import { UserHelperComponent } from '../user-helper/user-helper.component';
import { NotifyService } from '@farris/ui-notify';
import { I18nService } from '../../services/i18n.service';
export class TaskAssignComponent extends UserHelperComponent {
    //是否允许全员指派
    /**
     * @param {?} taskSvc
     * @param {?} userSvc
     * @param {?} notify
     * @param {?} i18nService
     */
    constructor(taskSvc, userSvc, notify, i18nService) {
        super(i18nService, userSvc);
        this.taskSvc = taskSvc;
        this.userSvc = userSvc;
        this.notify = notify;
        this.i18nService = i18nService;
        this.completed = new EventEmitter();
        this.paginationOptions.id = 'WF-TaskAss-DataTable-Pagination';
        // this.title = this.i18nService.getValue('static.assignee.title');
    }
    // title = '请选择指派用户';
    /**
     * @param {?} assigneeInfos
     * @return {?}
     */
    set assigneeInfo(assigneeInfos) {
        if (assigneeInfos && assigneeInfos.length > 0) {
            this.assignees = assigneeInfos;
            this.showModal(true);
            this.execute();
        }
        else {
            this.showModal(false);
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    execute() {
        this.assigneesPa = {
            workItemId: this.assignees[0].workItemId,
            assigneeIds: [],
            comment: ''
        };
        this.singleSelect = !this.assignees[0].multiSelect;
        this.allowGlobalAssign = this.assignees[0].allowGlobalAssign;
        if (this.assignees[0].assigneeUsers && this.assignees[0].assigneeUsers.length) {
            this.getAssignedUsers();
        }
        else {
            if (this.allowGlobalAssign) {
                this.getUserList();
            }
        }
    }
    /**
     * @return {?}
     */
    getAssignedUsers() {
        this.allUserData = this.assignees[0].assigneeUsers.map((/**
         * @param {?} user
         * @return {?}
         */
        user => {
            return {
                ID: user.id,
                Name: user.name,
                Code: user.code,
                OrgId: user.orgId,
                OrgName: user.orgNamePath
            };
        }));
        this.userListData = this.allUserData;
    }
    /**
     * @return {?}
     */
    handleOk() {
        this.selectedUsers = this.singleSelect ? [this.userTable.selections] : this.userTable.selections;
        if (!this.selectedUsers || this.selectedUsers.length < 1) {
            this.notify.warning(this.i18nService.getValue('static.assignee.noAssignee'));
            return;
        }
        this.assignees.shift();
        this.assigneesPa.assigneeIds = this.selectedUsers.map((/**
         * @param {?} user
         * @return {?}
         */
        user => user.ID));
        // 检查指定是否完成，如果完成提交，否则继续指派任务
        if (this.assignees.length > 0) {
            this.userListData = []; // clear source
            this.execute();
        }
        else {
            this.taskSvc.assignNextUsersByWorkItemId(this.assigneesPa).subscribe((/**
             * @return {?}
             */
            () => {
                this.showModal(false);
                this.notify.success(this.i18nService.getValue('static.assignee.successAssign'));
                this.completed.emit(true);
            }), (/**
             * @param {?} error
             * @return {?}
             */
            error => {
                if (error && error.error && error.error.Message) {
                    this.notify.error(error.error.Message);
                }
            }));
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    search(event) {
        if (!this.assignees[0].assigneeUsers || !this.assignees[0].assigneeUsers.length) {
            super.search(event);
        }
        else {
            this.searchValue = event.value;
            if (this.searchValue) {
                this.userListData = this.allUserData.filter((/**
                 * @param {?} user
                 * @return {?}
                 */
                user => (user.Name) && user.Name.includes(this.searchValue) ||
                    (user.Code) && user.Code.includes(this.searchValue)));
            }
            else {
                this.getAssignedUsers();
            }
        }
    }
    /**
     * @protected
     * @param {?} show
     * @return {?}
     */
    showModal(show) {
        if (show) {
            this.userDialog.show();
        }
        else {
            // 临时处理方案
            if (this.userDialog.beforeClose) {
                this.userDialog.close();
            }
        }
    }
    /**
     * @param {?} pageSize
     * @return {?}
     */
    onPageSizeChange(pageSize) {
        this.paginationOptions.itemsPerPage = pageSize;
        if (this.allowGlobalAssign) {
            this.getUserList();
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onPageChange(event) {
        this.paginationOptions.currentPage = event.pageIndex;
        if (this.allowGlobalAssign) {
            this.getUserList();
        }
    }
}
TaskAssignComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf-task-assign',
                template: "<farris-dialog #allUserDialog [buttons]=\"buttonRef\" [width]=\"850\" [height]=\"520\" [title]=\"'\u7CFB\u7EDF\u7528\u6237'\">\r\n  <div class=\"d-flex flex-column my-data-grid h-100\" style=\"padding: 0 14px;\">\r\n    <div class=\"flex-auto\" style=\"position:relative;\">\r\n      <farris-datatable #allUserTable [columns]=\"columns\" [data]=\"userListData | paginate: paginationOptions\"\r\n        [showFilterBar]='true' [singleSelect]=\"singleSelect\" [idField]=\"'ID'\" [remote]=\"'server'\" (search)=\"search($event)\"\r\n        [pagination]='false'></farris-datatable>\r\n    </div>\r\n    <div style=\"position:relative;\">\r\n      <pagination-controls [id]=\"paginationOptions.id\" [previousLabel]=\"' '\" [nextLabel]=\"' '\"\r\n        (pageChange)=\"onPageChange($event)\" (pageSizeChange)=\"onPageSizeChange($event)\">\r\n      </pagination-controls>\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n<ng-template #buttonRef>\r\n  <button class=\"btn btn-sm btn-primary px-3\" (click)=\"handleOk()\">\u786E\u5B9A</button>\r\n  <button class=\"btn btn-sm btn-default px-3\" (click)=\"handleCancel()\">\u53D6\u6D88</button>\r\n</ng-template>",
                styles: [".my-data-grid .filter-row .search-fields{padding-left:0!important}.my-data-grid .filter-row .search-text{padding-right:0!important}"]
            }] }
];
/** @nocollapse */
TaskAssignComponent.ctorParameters = () => [
    { type: TaskService },
    { type: UserService },
    { type: NotifyService },
    { type: I18nService }
];
TaskAssignComponent.propDecorators = {
    assigneeInfo: [{ type: Input }],
    completed: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    TaskAssignComponent.prototype.completed;
    /**
     * @type {?}
     * @private
     */
    TaskAssignComponent.prototype.allUserData;
    /**
     * @type {?}
     * @private
     */
    TaskAssignComponent.prototype.assignees;
    /**
     * @type {?}
     * @private
     */
    TaskAssignComponent.prototype.assigneesPa;
    /**
     * @type {?}
     * @private
     */
    TaskAssignComponent.prototype.allowGlobalAssign;
    /** @type {?} */
    TaskAssignComponent.prototype.taskSvc;
    /**
     * @type {?}
     * @protected
     */
    TaskAssignComponent.prototype.userSvc;
    /**
     * @type {?}
     * @protected
     */
    TaskAssignComponent.prototype.notify;
    /**
     * @type {?}
     * @protected
     */
    TaskAssignComponent.prototype.i18nService;
}
