// @gsp-lcm/bo-rt-service/esm2015/lib/business-object-rt.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/business-object-rt.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { of } from 'rxjs/observable/of';
import { HttpService } from '@ecp-caf/caf-common';
/** @type {?} */
const PROJECTS_DIRECTORY = '/api/dev/main/v1.0/business-object-trees';
/** @type {?} */
const url = '/api/runtime/sys/v1.0/business-objects/';
export class BusinessObjectRtService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @return {?}
     */
    getBusinessObjectTree() {
        return this.http.get(PROJECTS_DIRECTORY).pipe();
    }
    /**
     * @param {?} boId
     * @return {?}
     */
    getBusinessObject(boId) {
        /** @type {?} */
        const getUrl = url + boId;
        return this.http.get(getUrl).pipe();
    }
    /**
     * @return {?}
     */
    getBusinessObjects() {
        return this.http.get(url).pipe();
    }
    /**
     * 保存
     * @param {?} postData
     * @return {?}
     */
    save(postData) {
        return this.http.post(url, postData).pipe();
    }
    /**
     * @param {?} postData
     * @return {?}
     */
    saveChanges(postData) {
        return this.http.put(url, postData).pipe();
    }
    /**
     * 删除并重新加载
     * @param {?} dataId
     * @return {?}
     */
    remove(dataId) {
        /** @type {?} */
        const deleteUrl = url + dataId;
        return this.http.delete(deleteUrl, null).pipe();
    }
    /**
     * @param {?} boId
     * @return {?}
     */
    getDevBasicBOInfo(boId) {
        /** @type {?} */
        const devUrl = url + 'dev-basic-boinfos/';
        /** @type {?} */
        const getUrl = devUrl + boId;
        return this.http.get(getUrl).pipe();
    }
    /**
     * @param {?} parentId
     * @return {?}
     */
    getBOListByParentId(parentId) {
        /** @type {?} */
        const getUrl = url + '?parentID=' + parentId;
        return this.http.get(getUrl).pipe();
    }
    /**
     * @private
     * @template T
     * @param {?=} operation
     * @param {?=} result
     * @return {?}
     */
    handleError(operation = 'operation', result) {
        return (/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            console.error(error); // log to console instead
            return of((/** @type {?} */ (result)));
        });
    }
}
BusinessObjectRtService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
BusinessObjectRtService.ctorParameters = () => [
    { type: HttpService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    BusinessObjectRtService.prototype.http;
}
