// @edp-pmf/bpmn-model/dist/esm/lib/model/bpmndi/BPMNPlane.js
import { __extends } from "tslib";
import { BpmnModelHelper } from "../BpmnModelHelper";
import { ExtensionMethod } from "../ExtensionMethod";
import { Plane } from "../omgdi/Plane";
import { PalProcessAttribute } from "../mex/PalProcessAttribute";
var BPMNPlane = /** @class */ (function (_super) {
    __extends(BPMNPlane, _super);
    function BPMNPlane(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.bpmndi.BPMNPlane";
        return _this;
    }
    BPMNPlane.prototype.NewInstance = function () {
        return new BPMNPlane(this.Model);
    };
    BPMNPlane.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.BpmnElementId = null;
        this.palProcessAttribute = null;
    };
    BPMNPlane.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "BpmnElementId", this.BpmnElementId);
        return obj;
    };
    BPMNPlane.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.BpmnElementId = obj.BpmnElementId;
    };
    BPMNPlane.prototype.LoadExtensionElements = function (extensionElementsObj) {
        _super.prototype.LoadExtensionElements.call(this, extensionElementsObj);
        var palProcessAttributeObj = ExtensionMethod.GetAttributeValue(extensionElementsObj, "palProcessAttribute");
        if (palProcessAttributeObj != null) {
            this.palProcessAttribute = new PalProcessAttribute(this.Model);
            this.palProcessAttribute.LoadFromJsonObject(palProcessAttributeObj);
        }
    };
    BPMNPlane.prototype.WriteExtensionElements = function (extensionElementsObj) {
        _super.prototype.WriteExtensionElements.call(this, extensionElementsObj);
        if (this.palProcessAttribute != null) {
            ExtensionMethod.SetAttributeValue(extensionElementsObj, "palProcessAttribute", this.palProcessAttribute.ConvertToJsonObject());
        }
    };
    BPMNPlane.MEXTAG = "plane.bpa";
    return BPMNPlane;
}(Plane));
export { BPMNPlane };
