// @edp-pmf/bpmn-model/dist/esm/lib/model/bpmndi/BPMNLabel.js
import { __extends } from "tslib";
import { BpmnModelHelper } from "../BpmnModelHelper";
import { Label } from "../omgdi/Label";
var BPMNLabel = /** @class */ (function (_super) {
    __extends(BPMNLabel, _super);
    function BPMNLabel(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.bpmndi.BPMNLabel";
        return _this;
    }
    BPMNLabel.prototype.NewInstance = function () {
        return new BPMNLabel(this.Model);
    };
    BPMNLabel.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.LabelStyle = null;
    };
    BPMNLabel.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "LabelStyle", this.LabelStyle);
        return obj;
    };
    BPMNLabel.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.LabelStyle = obj.LabelStyle;
    };
    return BPMNLabel;
}(Label));
export { BPMNLabel };
