// @farris/devkit/esm5/lib/ui-state/ui_state.js
/*
 * @Author: Witt
 * @Date: 2018-11-17 13:38:23
 * @Last Modified by: Witt
 * @Last Modified time: 2018-11-17 13:38:50
 * @todo：临时删除原有功能，待重构
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { UIStateMetadataUtil } from './uistate_metadata_util';
/**
 * UI状态
 */
var UIState = /** @class */ (function () {
    function UIState() {
        this.paramTypeTransform = false;
        this.changes = new Subject();
        this.innerData = Object.assign({});
        this._init();
    }
    UIState.prototype._init = function () {
        var uiFields = UIStateMetadataUtil.getUIFields(this.constructor);
        this.initializeUIField(uiFields);
    };
    UIState.prototype.initialize = function (context) {
        var uiFields = context.metadata.uiStates || UIStateMetadataUtil.getUIFields(this.constructor);
        this.initializeUIField(uiFields);
    };
    UIState.prototype.initializeUIField = function (uiFieldMetadata) {
        var _this = this;
        Object.keys(uiFieldMetadata).forEach(function (propertyName) {
            var fieldMetadata = uiFieldMetadata[propertyName];
            var uiField = fieldMetadata.stateName || propertyName;
            if (delete _this[propertyName]) {
                _this.defineProperty(propertyName, uiField);
            }
        });
    };
    UIState.prototype.isExistProperty = function (propertyName) {
        if (this.innerData.hasOwnProperty(propertyName) || this.hasOwnProperty(propertyName)) {
            return true;
        }
        return false;
    };
    UIState.prototype.defineProperty = function (propertyName, field) {
        if (field === void 0) { field = null; }
        Object.defineProperty(this, propertyName, {
            get: function () {
                return field !== null ? this.innerData[field] : this.innerData[propertyName];
            },
            set: function (value) {
                // 值相同时不触发变更
                var oldValue = field !== null ? this.innerData[field] : this.innerData[propertyName];
                if (this.paramTypeTransform === true) {
                    var ngParams = UIStateMetadataUtil.getUIFields(this.constructor);
                    var ngParam = ngParams && ngParams[propertyName] || null;
                    var dataType = ngParam && ngParam.originalDataType || null;
                    if (dataType) {
                        value = this.transform(value, dataType);
                    }
                }
                if (oldValue === value) {
                    return;
                }
                if (field !== null) {
                    this.innerData[field] = value;
                }
                else {
                    this.innerData[propertyName] = value;
                }
                this.changes.next({
                    field: propertyName,
                    value: value
                });
            }
        });
    };
    UIState.prototype.setPropertyValue = function (propertyName, value) {
        if (propertyName === '' || propertyName === undefined) {
            return;
        }
        if (!this.isExistProperty(propertyName)) {
            this.defineProperty(propertyName);
        }
        this[propertyName] = value;
    };
    UIState.prototype.transform = function (target, dataType) {
        if (!dataType) {
            return target;
        }
        dataType = dataType.toLowerCase();
        if (dataType === 'string') {
            if (target === null || target === undefined) {
                return target;
            }
            return target.toString();
        }
        else if (dataType === 'number') {
            if (target === undefined) {
                return undefined;
            }
            var result = Number(target);
            if (isNaN(result)) {
                throw new Error(target + "\u65E0\u6CD5\u8F6C\u6362\u4E3A\u6570\u5B57\uFF01");
            }
            return result;
        }
        else if (dataType === 'boolean') {
            if (typeof target === 'boolean') {
                return target;
            }
            else {
                if (target === null || target === undefined) {
                    return false;
                }
                target = target.toString().toLowerCase();
                if (target === 'false') {
                    return false;
                }
                else if (target === 'true') {
                    return true;
                }
                else {
                    throw new Error(target + "\u65E0\u6CD5\u8F6C\u6362\u4E3A\u5E03\u5C14\u7C7B\u578B\uFF01");
                }
            }
        }
        else if (dataType === 'date' || dataType === 'datetime') {
            // 日期、日期时间在前端依然按照字符串处理
            return target;
        }
        else if (dataType === 'object') {
            if (typeof target === 'object') {
                return target;
            }
            else {
                try {
                    return JSON.parse(target);
                }
                catch (_a) {
                    throw new Error(target + "\u65E0\u6CD5\u8F6C\u6362\u4E3A\u5BF9\u8C61\uFF01");
                }
            }
        }
        else {
            return target;
        }
    };
    UIState.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    UIState.ctorParameters = function () { return []; };
    return UIState;
}());
export { UIState };
