// @farris/devkit/esm5/lib/event-mechanism/subscription/subscription.js
import { Injectable } from '@angular/core';
import { MetadataUtil } from '../../metadata/index';
import { NG_SUBSCRIPTION } from './subscription_decorator';
var Subscription = /** @class */ (function () {
    function Subscription() {
    }
    /**
     * 初始化
     */
    Subscription.prototype.init = function (frameComponent) {
        if (!frameComponent) {
            return;
        }
        return this.bindSubscription(frameComponent, null);
    };
    /**
     *  根据订阅列表进行初始化
     * @param frameComponent
     * @param ngEvents 订阅列表
     * @returns eventPipes
     */
    Subscription.prototype.initWithSubscriptions = function (frameComponent, ngEvents) {
        if (!frameComponent) {
            return;
        }
        return this.bindSubscription(frameComponent, ngEvents);
    };
    /**
     * 更改成为public形式 目的为了解析表单可以进行重新绑定
     * @param frameComponent
     * @param ngEvents
     * @returns
     */
    Subscription.prototype.bindSubscription = function (frameComponent, ngEvents) {
        var _this = this;
        var context = frameComponent.context;
        if (!context) {
            return;
        }
        var ngEventHandlerProps = ngEvents ? ngEvents : this.getNgEvents(frameComponent);
        if (!ngEventHandlerProps) {
            return;
        }
        var eventPipes = [];
        Object.keys(ngEventHandlerProps).forEach(function (propertyName) {
            var ngImportEvent = ngEventHandlerProps[propertyName];
            // 获取待订阅方法详情，尝试执行订阅
            var targetContext = context;
            var receiver = frameComponent;
            var emitter = ngImportEvent.token;
            var tokenValue = ngImportEvent.token;
            var eventName = ngImportEvent.name;
            var paramMapCollection = ngImportEvent.paramMapCollection;
            var eventPipe = targetContext.eventBus.on(emitter, tokenValue, eventName, receiver, function (eventArgs) {
                _this.subscriptionHandler(eventArgs, paramMapCollection, targetContext);
                var eventHandler = frameComponent[eventName];
                if (!eventHandler) {
                    return;
                }
                try {
                    eventHandler(receiver, eventArgs);
                }
                catch (_a) {
                    throw new Error('Error invoking method ' + eventName);
                }
            });
            eventPipes.push(eventPipe);
        });
        return eventPipes;
    };
    /**
     * 获取组件订阅列表
     * @param frameComponent 表单component
     * @returns 组件订阅列表信息
     */
    Subscription.prototype.getNgEvents = function (frameComponent) {
        return MetadataUtil.getPropsMetadatasByName(frameComponent.constructor, NG_SUBSCRIPTION);
    };
    Subscription.prototype.subscriptionHandler = function (param, paramMapCollection, currentFrameContext) {
        if (!param || !paramMapCollection || paramMapCollection.length <= 0 || !currentFrameContext) {
            return;
        }
        this.paramMap2UiState(param, paramMapCollection, currentFrameContext);
    };
    /**
     * 设置paramMap后，将param映射到UISTATE上
     */
    Subscription.prototype.paramMap2UiState = function (param, paramMapCollection, currentFrameContext) {
        for (var i = 0; i < paramMapCollection.length; i++) {
            var from = paramMapCollection[i].from;
            var frameId = paramMapCollection[i].frameId;
            var to = paramMapCollection[i].to;
            if (!from || !frameId || !to) {
                continue;
            }
            var destContext = this.getFrameContext(frameId, currentFrameContext);
            if (destContext == null) {
                continue;
            }
            this.setUiStateProperty(to, param[from], destContext.uiState);
            // this.setUiStateProperty(to, param[from], currentFrameContext.uiState);
        }
    };
    Subscription.prototype.getFrameContext = function (targetFrameContextId, currentContext) {
        var destContext = null;
        try {
            destContext = currentContext.appContext.getFrameContext(targetFrameContextId);
        }
        catch (_a) {
            throw new Error('Error in Getting FrameContext');
        }
        return destContext;
    };
    Subscription.prototype.setUiStateProperty = function (propertyName, propertyValue, uiState) {
        try {
            uiState.setPropertyValue(propertyName, propertyValue);
        }
        catch (_a) {
            throw new Error("Error in Setting Property Value of the current UISTATE" + uiState);
        }
    };
    Subscription.decorators = [
        { type: Injectable }
    ];
    return Subscription;
}());
export { Subscription };
