// @farris/ui-messager/esm5/lib/farris-default-focus.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, NgZone } from '@angular/core';
var FarrisDefaultFocusDirective = /** @class */ (function () {
    function FarrisDefaultFocusDirective(el, ngZone) {
        this.el = el;
        this.ngZone = ngZone;
    }
    /**
     * @return {?}
     */
    FarrisDefaultFocusDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    FarrisDefaultFocusDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.button.defaultFocus) {
            this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () {
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    _this.el.nativeElement.focus();
                }));
            }));
        }
    };
    FarrisDefaultFocusDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-default-focus]'
                },] }
    ];
    /** @nocollapse */
    FarrisDefaultFocusDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: NgZone }
    ]; };
    FarrisDefaultFocusDirective.propDecorators = {
        button: [{ type: Input, args: ['farris-default-focus',] }]
    };
    return FarrisDefaultFocusDirective;
}());
export { FarrisDefaultFocusDirective };
if (false) {
    /** @type {?} */
    FarrisDefaultFocusDirective.prototype.button;
    /**
     * @type {?}
     * @private
     */
    FarrisDefaultFocusDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FarrisDefaultFocusDirective.prototype.ngZone;
}
