// @farris/ui-common/esm5/lib/service/debug.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { fromEvent, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
var DebugService = /** @class */ (function () {
    function DebugService() {
        this.isDebug = false;
        this.destroy$ = new Subject();
        this.debugSub = null;
    }
    /**
     * @return {?}
     */
    DebugService.prototype.destroy = /**
     * @return {?}
     */
    function () {
        this.destroy$.next();
        this.destroy$.complete();
    };
    /**
     * @private
     * @param {?} type
     * @param {...?} msg
     * @return {?}
     */
    DebugService.prototype.writeMessage = /**
     * @private
     * @param {?} type
     * @param {...?} msg
     * @return {?}
     */
    function (type) {
        var msg = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            msg[_i - 1] = arguments[_i];
        }
        if (this.isDebug) {
            console[type].apply(console, tslib_1.__spread(msg));
        }
    };
    /**
     * @param {...?} msg
     * @return {?}
     */
    DebugService.prototype.log = /**
     * @param {...?} msg
     * @return {?}
     */
    function () {
        var msg = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            msg[_i] = arguments[_i];
        }
        this.writeMessage('log', msg);
    };
    /**
     * @param {...?} msg
     * @return {?}
     */
    DebugService.prototype.warn = /**
     * @param {...?} msg
     * @return {?}
     */
    function () {
        var msg = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            msg[_i] = arguments[_i];
        }
        this.writeMessage('warn', msg);
    };
    /**
     * @param {...?} msg
     * @return {?}
     */
    DebugService.prototype.error = /**
     * @param {...?} msg
     * @return {?}
     */
    function () {
        var msg = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            msg[_i] = arguments[_i];
        }
        this.writeMessage('error', msg);
    };
    /**
     * @param {...?} msg
     * @return {?}
     */
    DebugService.prototype.info = /**
     * @param {...?} msg
     * @return {?}
     */
    function () {
        var msg = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            msg[_i] = arguments[_i];
        }
        this.writeMessage('info', msg);
    };
    /**
     * @return {?}
     */
    DebugService.prototype.useDebugMode = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.debugSub) {
            this.debugSub = fromEvent(document.body, 'keydown').pipe(takeUntil(this.destroy$)).subscribe((/**
             * @param {?} e
             * @return {?}
             */
            function (e) {
                if (e.ctrlKey && e.shiftKey && e.key === 'D') {
                    e.preventDefault();
                    e.stopPropagation();
                    _this.isDebug = !_this.isDebug;
                    // console.log(`%c🪲 Farris debug mode is ${ this.isDebug ? 'startting' : 'stopped'}.🪲`, 'color: blue; font-weight:bold;font-size:16px')
                    console.log("%c\uD83E\uDEB2 Farris debug mode is " + (_this.isDebug ? 'startting' : 'stopped') + ".\uD83E\uDEB2", "color: " + (_this.isDebug ? 'green' : 'red') + "; font-weight:bold;font-size:16px");
                }
            }));
        }
        return this.debugSub;
    };
    DebugService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    DebugService.ctorParameters = function () { return []; };
    /** @nocollapse */ DebugService.ngInjectableDef = i0.defineInjectable({ factory: function DebugService_Factory() { return new DebugService(); }, token: DebugService, providedIn: "root" });
    return DebugService;
}());
export { DebugService };
if (false) {
    /** @type {?} */
    DebugService.prototype.isDebug;
    /**
     * @type {?}
     * @private
     */
    DebugService.prototype.destroy$;
    /**
     * @type {?}
     * @private
     */
    DebugService.prototype.debugSub;
}
