// @gspwidget/widget-core/esm2015/lib/flat-select/flat-select.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Input, Output, EventEmitter } from '@angular/core';
export class FlatSelectComponent {
    constructor() {
        this.options = [];
        this.valueChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
FlatSelectComponent.decorators = [
    { type: Component, args: [{
                selector: 'flat-select',
                template: "\n<ul class=\"d-flex\">\n  <li\n    *ngFor=\"let option of options\"\n    class=\"cursor-pointer transition\"\n    [class.active]=\"option.value === value\"\n    (click)=\"valueChange.emit(option.value)\"\n    >{{option.label || option.value}}</li>\n</ul>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: ["ul{margin-right:5px}ul li{padding:3px 10px;border-radius:100px}ul li.active,ul li:hover{background:#ecf0f3}"]
            }] }
];
/** @nocollapse */
FlatSelectComponent.ctorParameters = () => [];
FlatSelectComponent.propDecorators = {
    options: [{ type: Input }],
    value: [{ type: Input }],
    valueChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FlatSelectComponent.prototype.options;
    /** @type {?} */
    FlatSelectComponent.prototype.value;
    /** @type {?} */
    FlatSelectComponent.prototype.valueChange;
}
