// @qdp/search-join/fesm5/qdp-search-join.js
import { CommonModule } from '@angular/common';
import { __values } from 'tslib';
import { Injectable, RendererFactory2, Component, HostListener, ElementRef, Input, Optional, Injector, NgModule } from '@angular/core';
import { map } from 'rxjs/operators';
import { Subject, of } from 'rxjs';
import { FrameworkService, AppService } from '@gsp-sys/rtf-common';
import { CacheService } from '@ecp-caf/caf-common';
import { JointSearchManagerService, RtfServices, JointSearchExtendService, EventBus } from '@qdp/common';
import { LoadingService, LoadingModule } from '@farris/ui-loading';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SearchListService = /** @class */ (function () {
    function SearchListService(jointSearchManagerService) {
        this.jointSearchManagerService = jointSearchManagerService;
    }
    /**
     * 根据查询ID，获取联查信息
     * @param queryid 查询id
     */
    /**
     * 根据查询ID，获取联查信息
     * @param {?} formId
     * @param {?} queryid 查询id
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    SearchListService.prototype.getSearchList = /**
     * 根据查询ID，获取联查信息
     * @param {?} formId
     * @param {?} queryid 查询id
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (formId, queryid, queryRelativeUrl) {
        return this.jointSearchManagerService.getjointsearch(formId, queryid, queryRelativeUrl);
    };
    /**
     * @param {?} visibleJointParam
     * @param {?} queryRelativeUrl
     * @param {?} searchList
     * @return {?}
     */
    SearchListService.prototype.getVisibleJoint = /**
     * @param {?} visibleJointParam
     * @param {?} queryRelativeUrl
     * @param {?} searchList
     * @return {?}
     */
    function (visibleJointParam, queryRelativeUrl, searchList) {
        return this.jointSearchManagerService.getVisibleJoint(visibleJointParam, queryRelativeUrl, searchList);
    };
    SearchListService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SearchListService.ctorParameters = function () { return [
        { type: JointSearchManagerService }
    ]; };
    return SearchListService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MenuControllerService = /** @class */ (function () {
    function MenuControllerService(rendererFactory, cacheService, searchListService, cache, loadingService) {
        this.rendererFactory = rendererFactory;
        this.cacheService = cacheService;
        this.searchListService = searchListService;
        this.cache = cache;
        this.loadingService = loadingService;
        this.jointsearchInfoChange = new Subject();
        this.render = rendererFactory.createRenderer(null, null);
    }
    /**
     * @param {?} $event
     * @param {?} elementRef
     * @param {?} qdpView
     * @param {?} qdpViewElementRef
     * @return {?}
     */
    MenuControllerService.prototype.handleContextmenu = /**
     * @param {?} $event
     * @param {?} elementRef
     * @param {?} qdpView
     * @param {?} qdpViewElementRef
     * @return {?}
     */
    function ($event, elementRef, qdpView, qdpViewElementRef) {
        var _this = this;
        var e_1, _a;
        this.qdpView = qdpView;
        // const loading = this.loadingService.show({ container: qdpViewElementRef });
        /** @type {?} */
        var joinParam = this.cache.get(RtfServices.getTabId(this._queryId) + 'joinSearch');
        if (!this.checkDataArea(qdpView, qdpViewElementRef, joinParam, $event) || !this._searchList) {
            return false;
        }
        /** @type {?} */
        var left = 0;
        /** @type {?} */
        var top = 0;
        if (joinParam && joinParam.controlType !== 'charts') {
            left = $event.clientX;
            top = $event.clientY;
        }
        else {
            left = $event.event.event.clientX;
            top = $event.event.event.clientY;
        }
        this.render.setStyle(elementRef.nativeElement, 'left', left + 'px');
        this.render.setStyle(elementRef.nativeElement, 'top', top + 'px');
        this.render.setStyle(elementRef.nativeElement, 'display', 'block');
        this.cacheService.set(this._queryId + 'currentSelectItem', this.currentSelectItem); // 将当前参数设置到缓存中
        // 将当前参数设置到缓存中
        // 联查可见表达式
        /** @type {?} */
        var searchList = this.cacheService.get(this._queryId + 'beforeJointSearchMenuShow');
        /** @type {?} */
        var formulas = {};
        /** @type {?} */
        var containVisibleFormula = false;
        try {
            for (var searchList_1 = __values(searchList), searchList_1_1 = searchList_1.next(); !searchList_1_1.done; searchList_1_1 = searchList_1.next()) {
                var search = searchList_1_1.value;
                formulas[search.id] = search.visibleformula;
                if (containVisibleFormula === false && search.visibleformula != void 0 && search.visibleformula !== '') {
                    containVisibleFormula = true;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (searchList_1_1 && !searchList_1_1.done && (_a = searchList_1.return)) _a.call(searchList_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (containVisibleFormula === true) {
            /** @type {?} */
            var param = { qoManagerCode: qdpView.qoManagerCode, visibleFormulas: formulas, userId: '', queryId: this._queryId, rowData: this.currentSelectItem };
            this.searchListService.getVisibleJoint(param, qdpView.queryRelativeUrl, searchList).subscribe((/**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                if (value) {
                    _this.jointsearchInfoChange.next(value);
                }
            }));
        }
        else {
            this.jointsearchInfoChange.next(searchList);
        }
        return false;
    };
    /**
     * 单击之后隐藏菜单
     * @param elementRef 菜单组件
     */
    /**
     * 单击之后隐藏菜单
     * @param {?} elementRef 菜单组件
     * @return {?}
     */
    MenuControllerService.prototype.handleClick = /**
     * 单击之后隐藏菜单
     * @param {?} elementRef 菜单组件
     * @return {?}
     */
    function (elementRef) {
        if (elementRef) {
            this.render.setStyle(elementRef.nativeElement, 'display', 'none');
        }
    };
    /**
     * get searchList
     */
    /**
     * get searchList
     * @param {?} formId
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    MenuControllerService.prototype.getSearchList = /**
     * get searchList
     * @param {?} formId
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (formId, queryId, queryRelativeUrl) {
        var _this = this;
        this._queryId = queryId;
        if (this._searchList && this._searchList.queryId === this._queryId) {
            return of(this._searchList);
        }
        /** @type {?} */
        var result = this.searchListService.getSearchList(formId, queryId, queryRelativeUrl);
        return result.pipe(map((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this._searchList = value; // ? JSON.parse(value.joinSearch) : [];
            return _this._searchList;
        })));
    };
    /**
     * 返回当前右键坐标，是否为数据区域
     * @param qdpView qdp-view 组件
     * @param qdpViewElementRef qdp-view ElementRef
     * @param $event 触发事件，主要用来传递坐标
     */
    /**
     * 返回当前右键坐标，是否为数据区域
     * @private
     * @param {?} qdpView qdp-view 组件
     * @param {?} qdpViewElementRef qdp-view ElementRef
     * @param {?} joinParam
     * @param {?} $event 触发事件，主要用来传递坐标
     * @return {?}
     */
    MenuControllerService.prototype.checkDataArea = /**
     * 返回当前右键坐标，是否为数据区域
     * @private
     * @param {?} qdpView qdp-view 组件
     * @param {?} qdpViewElementRef qdp-view ElementRef
     * @param {?} joinParam
     * @param {?} $event 触发事件，主要用来传递坐标
     * @return {?}
     */
    function (qdpView, qdpViewElementRef, joinParam, $event) {
        var _this = this;
        /** @type {?} */
        var x = 0;
        /** @type {?} */
        var y = 0;
        /** @type {?} */
        var qdpviewNative = qdpViewElementRef.nativeElement;
        while (qdpviewNative) {
            x += qdpviewNative.offsetLeft;
            y += qdpviewNative.offsetTop;
            qdpviewNative = qdpviewNative.offsetParent;
        }
        if (joinParam.controlType !== 'charts') {
            /** @type {?} */
            var activeSheet = qdpView.spread.getActiveSheet();
            /** @type {?} */
            var target = activeSheet.hitTest($event.clientX - x, $event.clientY - y);
            if (target.hitTestType !== GC.Spread.Sheets.SheetArea.viewport || target.col === undefined || target.row === undefined) {
                return false;
            }
            activeSheet.setActiveCell(target.row, target.col);
            // 将当前行，写入cache
            /** @type {?} */
            var dataItem = {};
            if (joinParam.controlType === 'crosstab') { // 交叉表
                var _loop_1 = function (i) {
                    /** @type {?} */
                    var value_1 = activeSheet.getValue(target.row, i, GC.Spread.Sheets.SheetArea.viewport);
                    /** @type {?} */
                    var name_1 = activeSheet.getValue(joinParam.colHeaderRowInfo.start, i, GC.Spread.Sheets.SheetArea.colHeader);
                    /** @type {?} */
                    var colInfo_1 = joinParam.colInfos.find((/**
                     * @param {?} x
                     * @return {?}
                     */
                    function (x) { return x.dimension === 1 && x.name === name_1; })).bindField;
                    dataItem[colInfo_1] = value_1;
                };
                // 获取行标题的数据
                for (var i = joinParam.rowHeaderColInfo.start; i < joinParam.rowHeaderColInfo.end; i++) {
                    _loop_1(i);
                }
                // 获取列标题的数据
                for (var i = joinParam.colHeaderRowInfo.start; i < joinParam.colHeaderRowInfo.end; i++) {
                    /** @type {?} */
                    var value_2 = activeSheet.getValue(i, target.col, GC.Spread.Sheets.SheetArea.colHeader);
                    /** @type {?} */
                    var colInfo_2 = joinParam.colInfos.filter((/**
                     * @param {?} x
                     * @return {?}
                     */
                    function (x) { return x.dimension === 2; }))[i - joinParam.colHeaderRowInfo.start].bindField;
                    dataItem[colInfo_2] = value_2;
                }
                // 获取值标题的数据
                /** @type {?} */
                var value = activeSheet.getValue(target.row, target.col, GC.Spread.Sheets.SheetArea.viewport);
                /** @type {?} */
                var name_2 = activeSheet.getValue(joinParam.colHeaderRowInfo.end, target.col, GC.Spread.Sheets.SheetArea.colHeader);
                /** @type {?} */
                var colInfo = joinParam.colInfos.find((/**
                 * @param {?} x
                 * @return {?}
                 */
                function (x) { return x.dimension === 3 && x.name === name_2; }));
                if (colInfo) {
                    colInfo = colInfo.bindField;
                    dataItem[colInfo] = value;
                }
            }
            else {
                dataItem = activeSheet.getDataItem(target.row);
            }
            // 保存参数到searchList
            this.currentSelectItem = dataItem;
            return true;
        }
        else {
            if (joinParam.colList && joinParam.colList.length) {
                /** @type {?} */
                var cacheCols_1 = [];
                joinParam.colList.forEach((/**
                 * @param {?} col
                 * @return {?}
                 */
                function (col) {
                    _this.getColumn(col, null, cacheCols_1);
                }));
                /** @type {?} */
                var dataItem = {};
                for (var i = 0; i < cacheCols_1.length; i++) {
                    if (i < $event.value.length) {
                        dataItem[cacheCols_1[i].bindField] = $event.value[i];
                    }
                }
                this.currentSelectItem = dataItem;
            }
            $event.event.event.returnValue = false;
            return true;
        }
    };
    /**
     * 保存参数到cache
     * @param dataItem 参数
     */
    /**
     * 保存参数到cache
     * @protected
     * @param {?} dataItem 参数
     * @return {?}
     */
    MenuControllerService.prototype.writeParamToCache = /**
     * 保存参数到cache
     * @protected
     * @param {?} dataItem 参数
     * @return {?}
     */
    function (dataItem) {
        var _this = this;
        this._searchList[0]['joinSearch'].forEach((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            /** @type {?} */
            var currentParam = _this.cacheService.get(_this._queryId);
            dataItem['currentItem'] = currentParam;
            _this.cacheService.set(value.queryId, dataItem);
        }));
    };
    /**
     * @private
     * @param {?} col
     * @param {?} childCols
     * @param {?} colcache
     * @return {?}
     */
    MenuControllerService.prototype.getColumn = /**
     * @private
     * @param {?} col
     * @param {?} childCols
     * @param {?} colcache
     * @return {?}
     */
    function (col, childCols, colcache) {
        var _this = this;
        if (col) {
            /** @type {?} */
            var str = JSON.stringify(col);
            colcache.push(JSON.parse(str));
            if (col.childList && col.childList.length) {
                this.getColumn(null, col.childList, colcache);
            }
        }
        else {
            childCols.forEach((/**
             * @param {?} col
             * @return {?}
             */
            function (col) {
                /** @type {?} */
                var str = JSON.stringify(col);
                colcache.push(JSON.parse(str));
                if (col.childList && col.childList.length) {
                    _this.getColumn(null, col.childList, colcache);
                }
            }));
        }
    };
    MenuControllerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MenuControllerService.ctorParameters = function () { return [
        { type: RendererFactory2 },
        { type: CacheService },
        { type: SearchListService },
        { type: CacheService },
        { type: LoadingService }
    ]; };
    return MenuControllerService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SearchJoinMenuComponent = /** @class */ (function () {
    function SearchJoinMenuComponent(elementRef, renderFactory, menuControllerService, frameWorkService, cacheService, jointSearchExtendService, injector) {
        var _this = this;
        this.elementRef = elementRef;
        this.renderFactory = renderFactory;
        this.menuControllerService = menuControllerService;
        this.frameWorkService = frameWorkService;
        this.cacheService = cacheService;
        this.jointSearchExtendService = jointSearchExtendService;
        this.injector = injector;
        if (this.injector) {
            this.appService = this.injector.get(AppService);
        }
        this.formId = 'form' + 'ff3ea474-cc65-23b0-cdb1-73c9146cea3b';
        this.renderer = this.renderFactory.createRenderer(null, null);
        menuControllerService.jointsearchInfoChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value) {
                _this.searchList = value;
            }
        }));
    }
    /**
     * @return {?}
     */
    SearchJoinMenuComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var self = this;
        this.menuControllerService.getSearchList(this.formId, this.queryId, this.queryRelativeUrl).subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value) {
                /** @type {?} */
                var _searchList = JSON.parse(value['joinSearch']);
                _searchList.forEach((/**
                 * @param {?} searchL
                 * @return {?}
                 */
                function (searchL) {
                    if (!searchL.searchtype) {
                        searchL.searchtype = 'func';
                        searchL.appEntrance = '';
                    }
                    searchL.functionname = self.getFunctionName(searchL.functionname);
                }));
                _this.searchPara = JSON.parse(value['jointparam']);
                _this.cacheService.set(_this.queryId + 'beforeJointSearchMenuShow', _searchList);
            }
        }));
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    SearchJoinMenuComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
    };
    /**
     * 当点击某一个路由之后，隐藏菜单组件
     */
    /**
     * 当点击某一个路由之后，隐藏菜单组件
     * @return {?}
     */
    SearchJoinMenuComponent.prototype.handleClick = /**
     * 当点击某一个路由之后，隐藏菜单组件
     * @return {?}
     */
    function () {
        this.renderer.setStyle(this.elementRef.nativeElement, 'display', 'none');
    };
    /**
     * 路由到结果页
     * param id 联查功能主键id
     * param functionid 功能菜单id
     */
    /**
     * 路由到结果页
     * param id 联查功能主键id
     * param functionid 功能菜单id
     * @param {?} id
     * @param {?} functionid
     * @param {?} queryId
     * @return {?}
     */
    SearchJoinMenuComponent.prototype.routeTo = /**
     * 路由到结果页
     * param id 联查功能主键id
     * param functionid 功能菜单id
     * @param {?} id
     * @param {?} functionid
     * @param {?} queryId
     * @return {?}
     */
    function (id, functionid, queryId) {
        /** @type {?} */
        var self = this;
        /** @type {?} */
        var chooseList = {};
        this.searchList.forEach((/**
         * @param {?} searchL
         * @return {?}
         */
        function (searchL) {
            if (searchL.id === id) {
                chooseList = searchL;
            }
        }));
        /** @type {?} */
        var entityParam = { 'id': 'null' };
        /** @type {?} */
        var paramMap = new Map();
        /** @type {?} */
        var searchparas = this.searchPara.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.functionid === id; }));
        // 查找对应的参数列表
        /** @type {?} */
        var currentSelectedItem = this.cacheService.get(this.queryId + 'currentSelectItem');
        if (searchparas.length > 0) {
            searchparas.forEach((/**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                if (value.paravaluetype) { // 表达式
                    if (value.paravalue.indexOf('getParameter(') >= 0) {
                        /** @type {?} */
                        var parameterKey = value.paravalue.replace(/'/g, '').replace(/"/g, '').replace(/getParameter\(/g, '').replace(/\)/g, '');
                        self.getQueryParameters(entityParam, paramMap, parameterKey, value.para);
                    }
                    else {
                        entityParam[value.para] = currentSelectedItem[value.paravalue];
                        paramMap.set(value.para, currentSelectedItem[value.paravalue]);
                    }
                }
                else {
                    entityParam[value.para] = value.paravalue;
                    paramMap.set(value.para, value.paravalue);
                }
            }));
        }
        else {
            entityParam = currentSelectedItem;
        }
        /** @type {?} */
        var options = {
            ParentSessionId: this.cacheService.get('session'),
            EntityParam: entityParam
        };
        /** @type {?} */
        var $event = {};
        if (this.jointSearchExtendService && this.jointSearchExtendService.beforeJointSearch) {
            $event['eventCode'] = 'beforeJointSearch';
            $event['eventName'] = 'drill before';
            $event['data'] = {
                'functionid': functionid,
                'options': options,
                'queryId': queryId,
                'orginQueryId': this.queryId,
                'qdpView': this.menuControllerService.qdpView
            };
            this.jointSearchExtendService.beforeJointSearch($event);
            if ($event && $event['data'] && $event['data'].options) {
                options = Object.assign(options, $event['data'].options);
            }
            if (chooseList.searchtype === 'url' || chooseList.searchtype === 'additional') {
                return;
            }
        }
        // this.cacheService.set(queryId, options.EntityParam);
        /** @type {?} */
        var tabId = '';
        if (chooseList.searchtype === 'func') {
            // 联查前
            /** @type {?} */
            var openMenuoptions = {
                'appType': 'menu',
                'funcId': functionid,
                'tabId': EventBus.guid(),
                'appId': '',
                'appEntrance': '',
                'entityParams': $event['entityParams'] || entityParam,
                'queryStringParams': paramMap,
                'isNewTab': true
            };
            tabId = RtfServices.getFuncId(openMenuoptions);
            /** @type {?} */
            var menuSwitchControl = Object.assign({}, openMenuoptions['entityParams'], RtfServices.getMenuSwitchControlParameter(tabId));
            openMenuoptions['entityParams'] = menuSwitchControl;
            this.cacheService.set(tabId ? tabId : queryId, options.EntityParam);
            // 卡片塞进去params
            this.frameWorkService.openMenu(openMenuoptions);
        }
        else if (chooseList.searchtype === 'app') {
            /** @type {?} */
            var appOption = {
                appType: 'app',
                appId: functionid,
                appEntrance: chooseList.appEntrance,
                funcId: '',
                isReload: true,
                tabId: EventBus.guid(),
                entityParams: $event['entityParams'] || entityParam,
                queryStringParams: paramMap
            };
            tabId = RtfServices.getFuncId(appOption);
            this.cacheService.set(tabId ? tabId : queryId, options.EntityParam);
            /** @type {?} */
            var menuSwitchControl = Object.assign({}, appOption['entityParams'], RtfServices.getMenuSwitchControlParameter(tabId));
            appOption['entityParams'] = menuSwitchControl;
            this.frameWorkService.openMenu(appOption);
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    SearchJoinMenuComponent.prototype.getFunctionName = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        /** @type {?} */
        var local = 'zh-chs';
        if (gspframeworkService) {
            local = gspframeworkService.rtf.language.getLanguageCode().toLowerCase();
        }
        if (!local) {
            local = 'zh-chs';
        }
        return data[local] ? data[local] : data;
    };
    /**
     * @private
     * @param {?} entityParam
     * @param {?} paramMap
     * @param {?} parameterKey
     * @param {?} key
     * @return {?}
     */
    SearchJoinMenuComponent.prototype.getQueryParameters = /**
     * @private
     * @param {?} entityParam
     * @param {?} paramMap
     * @param {?} parameterKey
     * @param {?} key
     * @return {?}
     */
    function (entityParam, paramMap, parameterKey, key) {
        try {
            /** @type {?} */
            var tab = gspframeworkService.rtf.session.getCommonVariable();
            if (tab && tab.tabId) {
                /** @type {?} */
                var tid = tab.tabId + RtfServices.getInSuiteFrmUUID();
                if (this.cacheService.get(tid) && this.cacheService.get(tid)[parameterKey]) {
                    entityParam[key] = this.cacheService.get(tid)[parameterKey];
                    paramMap.set(key, this.cacheService.get(tid)[parameterKey]);
                }
            }
        }
        catch (e) {
        }
    };
    SearchJoinMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'qdp-search-join-menu',
                    template: "<div *ngIf=\"searchList && searchList.length && searchList.length > 0\">\r\n  <div class=\"dropdown-menu\" style=\"display:block\">\r\n    <div *ngFor=\"let item of searchList\">\r\n      <button class=\"dropdown-item\" (click)=\"routeTo(item.id,item.functioncode,item.destinationqueryId)\">{{item.functionname}}</button>\r\n    </div>\r\n  </div>\r\n</div>\r\n",
                    styles: [":host{position:fixed;display:none;top:0;left:0}"]
                }] }
    ];
    /** @nocollapse */
    SearchJoinMenuComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: RendererFactory2 },
        { type: MenuControllerService },
        { type: FrameworkService },
        { type: CacheService },
        { type: JointSearchExtendService, decorators: [{ type: Optional }] },
        { type: Injector, decorators: [{ type: Optional }] }
    ]; };
    SearchJoinMenuComponent.propDecorators = {
        queryId: [{ type: Input }],
        queryRelativeUrl: [{ type: Input }],
        qoManagerCode: [{ type: Input }],
        searchList: [{ type: Input }],
        handleClick: [{ type: HostListener, args: ['click',] }]
    };
    return SearchJoinMenuComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SearchJoinModule = /** @class */ (function () {
    function SearchJoinModule() {
    }
    SearchJoinModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        LoadingModule,
                    ],
                    providers: [
                        SearchListService,
                        LoadingService
                    ],
                    declarations: [
                        SearchJoinMenuComponent
                    ],
                    exports: [
                        SearchJoinMenuComponent
                    ]
                },] }
    ];
    return SearchJoinModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { SearchJoinModule, MenuControllerService, SearchListService, SearchJoinMenuComponent };
