// @farris/command-services/esm2015/lib/data-services/utils/edit-state.util.js
const IS_ADD = 'IS_ADD';
const LAST_MODIFIED_ID = 'LAST_MODIFIED_ID';
/**
 * 表单状态工具类
 * @internal
 */
class EditStateUtil {
    /**
     * 获取当前行id
     */
    static getLastModifiedId(frameContext) {
        // const lastModifiedId = AppParamUtil.get(frameContext, LAST_MODIFIED_ID);
        const virtualRootFrameContext = frameContext.getVirtualRootFrameContext();
        let lastModifiedId = null;
        if (virtualRootFrameContext) {
            lastModifiedId = virtualRootFrameContext.getParam(LAST_MODIFIED_ID);
            return lastModifiedId ? lastModifiedId : null;
        }
        return lastModifiedId;
    }
    /**
     * 设置最后修改的id
     */
    static setLastModifiedId(frameContext, lastModifiedId) {
        // AppParamUtil.set(frameContext, LAST_MODIFIED_ID, lastModifiedId);
        const virtualRootFrameContext = frameContext.getVirtualRootFrameContext();
        if (virtualRootFrameContext) {
            virtualRootFrameContext.setParam(LAST_MODIFIED_ID, lastModifiedId);
        }
    }
    /**
     * 获取新增状态
     */
    static getAddState(frameContext) {
        // return AppParamUtil.get(frameContext, IS_ADD);
        const virtualRootFrameContext = frameContext.getVirtualRootFrameContext();
        if (virtualRootFrameContext) {
            return virtualRootFrameContext.getParam(IS_ADD);
        }
    }
    /**
     * 设置新增状态
     */
    static setAddState(frameContext, lastModifiedId) {
        // AppParamUtil.set(frameContext, IS_ADD, true);
        const virtualRootFrameContext = frameContext.getVirtualRootFrameContext();
        if (virtualRootFrameContext) {
            virtualRootFrameContext.setParam(IS_ADD, true);
        }
        this.setLastModifiedId(frameContext, lastModifiedId);
    }
    /**
     * 取消新增状态
     */
    static cancelAddState(frameContext) {
        // AppParamUtil.set(frameContext, IS_ADD, false);
        const virtualRootFrameContext = frameContext.getVirtualRootFrameContext();
        if (virtualRootFrameContext) {
            virtualRootFrameContext.setParam(IS_ADD, false);
        }
    }
    /**
     * 设置编辑状态
     */
    static setEditState(frameContext, id) {
        // AppParamUtil.set(frameContext, IS_ADD, false);
        const virtualRootFrameContext = frameContext.getVirtualRootFrameContext();
        if (virtualRootFrameContext) {
            virtualRootFrameContext.setParam(IS_ADD, false);
        }
        this.setLastModifiedId(frameContext, id);
    }
}
export { EditStateUtil };
