// @gsp-svc/formdoc-upload/esm5/lib/uploaddialog/uploaddialog/uploaddialog.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Input } from '@angular/core';
import { BsModalService } from '@farris/ui-modal';
import { FileUploadComponent } from '../../upload/file-upload/file-upload.component';
var UploadDialogComponent = /** @class */ (function () {
    function UploadDialogComponent(modalService) {
        this.modalService = modalService;
        this.metadataIdList = [];
    }
    /**
     * @return {?}
     */
    UploadDialogComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    UploadDialogComponent.prototype.showDialog = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.dlg = this.modalService.show(FileUploadComponent, {
            title: '选择文件',
            width: 800, height: 500,
            buttons: [
                {
                    text: '上传', cls: 'k-button k-button-icontext k-primary', iconCls: "k-icon k-i-check", handle: (/**
                     * @return {?}
                     */
                    function () {
                        _this.fileUpload.upload();
                        _this.dlg.close();
                    })
                },
                {
                    text: '关闭', cls: 'k-button k-button-icontext', iconCls: "k-icon k-i-close", handle: (/**
                     * @return {?}
                     */
                    function () {
                        _this.fileUpload.cancel();
                        _this.dlg.close();
                    })
                }
            ],
            showButtons: true,
        });
        this.dlg.content.formId = this.formId;
    };
    UploadDialogComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-uploaddialog',
                    template: "\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    UploadDialogComponent.ctorParameters = function () { return [
        { type: BsModalService }
    ]; };
    UploadDialogComponent.propDecorators = {
        formId: [{ type: Input }],
        fileUpload: [{ type: ViewChild, args: [FileUploadComponent,] }]
    };
    return UploadDialogComponent;
}());
export { UploadDialogComponent };
if (false) {
    /** @type {?} */
    UploadDialogComponent.prototype.dlg;
    /** @type {?} */
    UploadDialogComponent.prototype.metadataIdList;
    /** @type {?} */
    UploadDialogComponent.prototype.formId;
    /** @type {?} */
    UploadDialogComponent.prototype.fileUpload;
    /**
     * @type {?}
     * @private
     */
    UploadDialogComponent.prototype.modalService;
}
