// @gsp-aif/common-api/fesm2015/gsp-aif-common-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const AifCallType = {
    /**
     * 默认值
     */
    Default: 0,
    /**
     * 同步调用
     */
    Sync: 1,
    /**
     * 异步调用
     */
    Async: 2,
};
AifCallType[AifCallType.Default] = 'Default';
AifCallType[AifCallType.Sync] = 'Sync';
AifCallType[AifCallType.Async] = 'Async';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const AifBizActionType = {
    /**
     * 生单
     */
    Creation: 1,
    /**
     * 回写
     */
    Writeback: 2,
};
AifBizActionType[AifBizActionType.Creation] = 'Creation';
AifBizActionType[AifBizActionType.Writeback] = 'Writeback';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 实体抽象类
 * @abstract
 */
class EntityBase {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 元数据实体抽象类
 * @abstract
 */
class MetadataEntityBase extends EntityBase {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 业务动作实体类
 */
class AifBizActionEntity extends MetadataEntityBase {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 业务场景图实体类
 */
class AifBizFlowchartEntity extends MetadataEntityBase {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 业务事件实体类
 */
class AifBizEventEntity extends MetadataEntityBase {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 点到点元数据实体抽象类
 * @abstract
 */
class P2PMetadataEntityBase extends MetadataEntityBase {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 映射规则实体抽象类
 * @abstract
 */
class MappingEntity extends P2PMetadataEntityBase {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 生单规则实体类
 */
class AifCreationRuleEntity extends MappingEntity {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 回写规则实体类
 */
class AifWritebackRuleEntity extends MappingEntity {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 业务流程实体抽象类
 * @abstract
 */
class BizProcessEntity extends P2PMetadataEntityBase {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 生单流程实体类
 */
class AifCreationProcessEntity extends BizProcessEntity {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 回写流程实体类
 */
class AifWritebackProcessEntity extends BizProcessEntity {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { AifCallType, AifBizActionType, AifBizActionEntity, AifBizFlowchartEntity, AifBizEventEntity, AifCreationRuleEntity, AifWritebackRuleEntity, AifCreationProcessEntity, AifWritebackProcessEntity, BizProcessEntity as ɵe, EntityBase as ɵb, MappingEntity as ɵc, MetadataEntityBase as ɵa, P2PMetadataEntityBase as ɵd };
