// @farris/ui-filter-panel/esm2015/lib/components/radiocomponent/radio.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output } from '@angular/core';
export class PanelRadioComponent {
    constructor() {
        this.searchHandle = new EventEmitter();
        /**
         * 判断是否是在扩展区域展示
         */
        this.extend = false;
        /**
         * 判断是否显示label
         */
        this.showlabel = true;
        this.selectCheck = [];
        this.radioDataSource = new Array();
    }
    /**
     * @return {?}
     */
    get item() {
        return this._item;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    set item(data) {
        this._item = data;
        this.getDataSource();
        if (!this._item.hasOwnProperty('value')) {
            this._item.value = { value: '' };
        }
        //this.radioValue = this._item.value.getOriginalValue();
        this.radioValue = this._item.value.value;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.getDataSource();
    }
    /**
     * @return {?}
     */
    ngOnChanges() {
        this.getDataSource();
    }
    /**
     * @return {?}
     */
    getDataSource() {
        if (this._item && this._item.control) {
            /** @type {?} */
            let radioboxList = (this._item.control);
            this.radioDataSource = Array();
            if (radioboxList.enumValues && radioboxList.enumValues.length > 0) {
                radioboxList.enumValues.forEach((/**
                 * @param {?} element
                 * @return {?}
                 */
                element => {
                    this.radioDataSource.push(Object.assign({}, element));
                }));
            }
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onSearch(event) {
        if (event) {
            this._item.value.value = event;
        }
        else {
            this._item.value.value = '';
        }
        // this._item.valueText = this._item.value.value;
        this.getValueText();
        //this.radioValue = this._item.value.getOriginalValue();
        this.radioValue = this._item.value.value;
        this.searchHandle.emit();
    }
    /**
     * @return {?}
     */
    getValueText() {
        /** @type {?} */
        const enumValue = (this._item.control).enumValues;
        /** @type {?} */
        const value = (this._item.value).value;
        if (enumValue && enumValue.length) {
            /** @type {?} */
            const texts = enumValue.find((/**
             * @param {?} val
             * @return {?}
             */
            val => {
                return val['value'] == value;
            }));
            if (texts) {
                this._item.valueText = texts['name'];
            }
        }
    }
}
PanelRadioComponent.decorators = [
    { type: Component, args: [{
                selector: 'panel-filter-radio',
                template: "<!-- <farris-radiogroup \r\n[value]=\"radioValue\" \r\n[horizontal]=\"false\" \r\n[data]=\"radioDataSource\"\r\n(changeValue)=\"onSearch($event)\">\r\n</farris-radiogroup> -->\r\n\r\n<div class=\"f-filter-panel-radiogroup\">\r\n    <div class=\"panel-radiogroup-item\" \r\n    [ngClass]=\"{'panel-radiogroup-item-active': radioValue == item.value}\"\r\n    *ngFor=\"let item of radioDataSource\"\r\n    (click)=\"onSearch(item.value)\"\r\n    >\r\n        <span class=\"panel-radiogroup-item-text\">{{item.name}}</span>\r\n    </div>\r\n</div>"
                // encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
PanelRadioComponent.ctorParameters = () => [];
PanelRadioComponent.propDecorators = {
    item: [{ type: Input }],
    searchHandle: [{ type: Output, args: ['searchHandle',] }]
};
if (false) {
    /** @type {?} */
    PanelRadioComponent.prototype._item;
    /** @type {?} */
    PanelRadioComponent.prototype.searchHandle;
    /**
     * 判断是否是在扩展区域展示
     * @type {?}
     */
    PanelRadioComponent.prototype.extend;
    /**
     * 判断是否显示label
     * @type {?}
     */
    PanelRadioComponent.prototype.showlabel;
    /** @type {?} */
    PanelRadioComponent.prototype.radioValue;
    /** @type {?} */
    PanelRadioComponent.prototype.selectCheck;
    /** @type {?} */
    PanelRadioComponent.prototype.radioDataSource;
}
