// @farris/ide-devkit/esm2015/lib/components/angular-editor/editor.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ANGULAR_MONACO_EDITOR_CONFIG } from './model/config';
import { AngularMonacoEditorComponent } from './angular-monaco-editor/angular-monaco-editor.component';
export class AngularMonacoEditorModule {
    /**
     * @param {?=} config
     * @return {?}
     */
    static forRoot(config = {}) {
        return {
            ngModule: AngularMonacoEditorModule,
            providers: [
                { provide: ANGULAR_MONACO_EDITOR_CONFIG, useValue: config }
            ]
        };
    }
}
AngularMonacoEditorModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    AngularMonacoEditorComponent
                ],
                exports: [
                    AngularMonacoEditorComponent
                ],
                providers: []
            },] }
];
