// @farris/extend-file-upload/esm5/lib/components/upload-and-preview/file-order-drag.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/upload-and-preview/file-order-drag.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, NgZone, Injector, Renderer2, ElementRef, Input, Output, EventEmitter, } from "@angular/core";
import { dropHandlers, smoothDnD as SmoothDndForVisible, constants, } from "@farris/smooth-dnd";
var wrapperClass = constants.wrapperClass, animationClass = constants.animationClass;
SmoothDndForVisible.dropHandler = dropHandlers.reactDropHandler().handler;
SmoothDndForVisible.wrapChild = false;
/**
 * @record
 */
export function DragStartEndInfo() { }
if (false) {
    /** @type {?} */
    DragStartEndInfo.prototype.isSource;
    /** @type {?} */
    DragStartEndInfo.prototype.payload;
    /** @type {?} */
    DragStartEndInfo.prototype.willAcceptDrop;
}
var UploadFileOrderDragDirective = /** @class */ (function () {
    function UploadFileOrderDragDirective(ngzone, injector, render, el) {
        var _this = this;
        this.ngzone = ngzone;
        this.injector = injector;
        this.render = render;
        this.el = el;
        this.enableDrag = false;
        this.drop = new EventEmitter();
        this.options = {
            behaviour: "move",
            dragHandleSelector: ".upfile-drag-handle",
            dropPlaceholder: {
                className: "drop-group-field",
            },
            getGhostParent: (/**
             * @return {?}
             */
            function () {
                return document.body;
            }),
            getChildPayload: this.getChildPayload.bind(this),
            shouldAcceptDrop: (/**
             * @param {?} sourceContainerOptions
             * @param {?} payload
             * @return {?}
             */
            function (sourceContainerOptions, payload) {
                return true;
            }),
            onDropReady: (/**
             * @param {?} dropResult
             * @return {?}
             */
            function (dropResult) {
                _this.ngzone.run((/**
                 * @return {?}
                 */
                function () {
                    _this.onDropReady(dropResult);
                }));
            }),
            onDrop: (/**
             * @param {?} dropResult
             * @return {?}
             */
            function (dropResult) {
                _this.ngzone.run((/**
                 * @return {?}
                 */
                function () {
                    _this.drop.emit(dropResult);
                }));
            }),
            onDragEnter: (/**
             * @return {?}
             */
            function () {
                _this.ngzone.run((/**
                 * @return {?}
                 */
                function () {
                    _this.onDragEnter();
                }));
            }),
            onDragStart: (/**
             * @param {?} info
             * @return {?}
             */
            function (info) {
                _this.ngzone.run((/**
                 * @return {?}
                 */
                function () {
                    _this.onDragStart(info);
                }));
            }),
            onDragEnd: (/**
             * @param {?} info
             * @return {?}
             */
            function (info) {
                _this.ngzone.run((/**
                 * @return {?}
                 */
                function () {
                    _this.onDragEnd(info);
                }));
            }),
        };
    }
    /**
     * @return {?}
     */
    UploadFileOrderDragDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.enableDrag) {
            this.initDnD();
        }
    };
    /**
     * @return {?}
     */
    UploadFileOrderDragDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.disposeDnd();
    };
    /**
     * @private
     * @return {?}
     */
    UploadFileOrderDragDirective.prototype.disposeDnd = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.container) {
            this.container.dispose();
            this.container = null;
        }
    };
    /**
     * @private
     * @return {?}
     */
    UploadFileOrderDragDirective.prototype.initDnD = /**
     * @private
     * @return {?}
     */
    function () {
        this.disposeDnd();
        this.container = SmoothDndForVisible(this.el.nativeElement, this.options);
    };
    /**
     * @private
     * @param {?} index
     * @return {?}
     */
    UploadFileOrderDragDirective.prototype.getChildPayload = /**
     * @private
     * @param {?} index
     * @return {?}
     */
    function (index) {
        return index;
    };
    /**
     * @private
     * @param {?} dropResult
     * @return {?}
     */
    UploadFileOrderDragDirective.prototype.onDropReady = /**
     * @private
     * @param {?} dropResult
     * @return {?}
     */
    function (dropResult) {
        // console.log('DROP READY', dropResult);
    };
    /**
     * @private
     * @param {?} info
     * @return {?}
     */
    UploadFileOrderDragDirective.prototype.onDragStart = /**
     * @private
     * @param {?} info
     * @return {?}
     */
    function (info) {
        // console.log("DRAG START", info);
    };
    /**
     * @private
     * @param {?} info
     * @return {?}
     */
    UploadFileOrderDragDirective.prototype.onDragEnd = /**
     * @private
     * @param {?} info
     * @return {?}
     */
    function (info) {
        // console.log('DRAG END', info);
    };
    /**
     * @private
     * @return {?}
     */
    UploadFileOrderDragDirective.prototype.onDragEnter = /**
     * @private
     * @return {?}
     */
    function () {
        // console.log('DRAG ENTER');
    };
    UploadFileOrderDragDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[upfile-order-drag]",
                },] }
    ];
    /** @nocollapse */
    UploadFileOrderDragDirective.ctorParameters = function () { return [
        { type: NgZone },
        { type: Injector },
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    UploadFileOrderDragDirective.propDecorators = {
        enableDrag: [{ type: Input, args: ['upfile-order-drag',] }],
        drop: [{ type: Output }],
        options: [{ type: Input }]
    };
    return UploadFileOrderDragDirective;
}());
export { UploadFileOrderDragDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    UploadFileOrderDragDirective.prototype.container;
    /** @type {?} */
    UploadFileOrderDragDirective.prototype.enableDrag;
    /** @type {?} */
    UploadFileOrderDragDirective.prototype.drop;
    /** @type {?} */
    UploadFileOrderDragDirective.prototype.options;
    /**
     * @type {?}
     * @private
     */
    UploadFileOrderDragDirective.prototype.ngzone;
    /**
     * @type {?}
     * @private
     */
    UploadFileOrderDragDirective.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    UploadFileOrderDragDirective.prototype.render;
    /**
     * @type {?}
     * @private
     */
    UploadFileOrderDragDirective.prototype.el;
}
