// @farris/ui-treetable/esm2015/lib/high-light.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe, ElementRef } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
export class HighlightPipe {
    /**
     * @param {?} sanitizer
     * @param {?} el
     */
    constructor(sanitizer, el) {
        this.sanitizer = sanitizer;
        this.el = el;
    }
    /**
     * @param {?} text
     * @param {?} keyword
     * @param {?=} field
     * @param {?=} findField
     * @return {?}
     */
    transform(text, keyword, field = '', findField = '') {
        /** @type {?} */
        let needSearch = true;
        if (field && findField) {
            if (findField.indexOf(',') === -1) {
                needSearch = field === findField;
            }
            else {
                needSearch = findField.split(',').some((/**
                 * @param {?} n
                 * @return {?}
                 */
                n => n === field));
            }
        }
        if (!needSearch) {
            return text;
        }
        if (keyword && text) {
            /** @type {?} */
            let pattern = keyword.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, '\\$&');
            pattern = pattern.split(' ').filter((/**
             * @param {?} t
             * @return {?}
             */
            (t) => {
                return t.length > 0;
            })).join('|');
            /** @type {?} */
            const regex = new RegExp(pattern, 'gi');
            return this.sanitizer.bypassSecurityTrustHtml(
            // text.replace(regex, (match) => `<span class="search-highlight">${match}</span>`)
            this.replaceContent(text, regex, (/**
             * @param {?} match
             * @return {?}
             */
            (match) => `<span class="search-highlight">${match}</span>`)));
        }
        else {
            return text;
        }
    }
    /**
     * @param {?} html
     * @param {?} context
     * @return {?}
     */
    parseHTML(html, context) {
        /** @type {?} */
        const t = (context || document).createElement('template');
        t.innerHTML = html;
        return t.content.cloneNode(true);
    }
    /**
     * @private
     * @param {?} strMatch1
     * @param {?} regex
     * @param {?} match
     * @return {?}
     */
    replaceContent(strMatch1, regex, match) {
        /** @type {?} */
        const matchReg1 = />.*?\</gi;
        // 先查出><之间的内容
        /** @type {?} */
        const strs = strMatch1.match(matchReg1);
        if (strs && strs.length) {
            for (const aa of strs) {
                /** @type {?} */
                const rep_old = strs[aa];
                // 拿到原始符合字符串
                /** @type {?} */
                const rep_new = strs[aa].replace(regex, match);
                // 替换成目的字符串
                /** @type {?} */
                const re2 = new RegExp(rep_old, 'g');
                strMatch1 = strMatch1.replace(re2, rep_new); // 将符合的原始字符串替换成新字符串
            }
            return strMatch1;
        }
        else {
            return strMatch1.replace(regex, match);
        }
    }
}
HighlightPipe.decorators = [
    { type: Pipe, args: [{ name: 'highlight' },] }
];
/** @nocollapse */
HighlightPipe.ctorParameters = () => [
    { type: DomSanitizer },
    { type: ElementRef }
];
if (false) {
    /** @type {?} */
    HighlightPipe.prototype.sanitizer;
    /**
     * @type {?}
     * @private
     */
    HighlightPipe.prototype.el;
}
