// @farris/ui-treetable/esm2015/lib/hotkeys.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { EventManager } from '@angular/platform-browser';
import { Observable } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
/**
 * @record
 */
function HotkeyOptions() { }
if (false) {
    /** @type {?} */
    HotkeyOptions.prototype.element;
    /** @type {?} */
    HotkeyOptions.prototype.description;
    /** @type {?} */
    HotkeyOptions.prototype.keys;
}
export class TTHotkeys {
    /**
     * @param {?} eventManager
     */
    constructor(eventManager) {
        this.eventManager = eventManager;
        this.hotkeys = new Map();
        this.defaults = {
            element: document
        };
    }
    /**
     * @param {?} options
     * @return {?}
     */
    addShortcut(options) {
        /** @type {?} */
        const merged = Object.assign({}, this.defaults, options);
        if (merged.keys.indexOf('ctrl') > -1) {
            merged.keys = merged.keys.replace('ctrl', 'control');
        }
        /** @type {?} */
        const event = `keydown.${merged.keys}`;
        if (merged.description) {
            this.hotkeys.set(merged.keys, merged.description);
        }
        return new Observable((/**
         * @param {?} observer
         * @return {?}
         */
        observer => {
            /** @type {?} */
            const handler = (/**
             * @param {?} e
             * @return {?}
             */
            (e) => {
                e.preventDefault();
                observer.next(e);
            });
            /** @type {?} */
            const dispose = this.eventManager.addEventListener(merged.element, event, handler);
            return (/**
             * @return {?}
             */
            () => {
                dispose();
                this.hotkeys.delete(merged.keys);
            });
        }));
    }
}
TTHotkeys.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TTHotkeys.ctorParameters = () => [
    { type: EventManager }
];
/** @nocollapse */ TTHotkeys.ngInjectableDef = i0.defineInjectable({ factory: function TTHotkeys_Factory() { return new TTHotkeys(i0.inject(i1.EventManager)); }, token: TTHotkeys, providedIn: "root" });
if (false) {
    /** @type {?} */
    TTHotkeys.prototype.hotkeys;
    /** @type {?} */
    TTHotkeys.prototype.defaults;
    /**
     * @type {?}
     * @private
     */
    TTHotkeys.prototype.eventManager;
}
