// @farris/ui-treetable/esm2015/lib/virtualized-loader.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class VirtualizedLoaderService {
    /**
     * @param {?} tt
     */
    constructor(tt) {
        this.tt = tt;
    }
    /**
     * @return {?}
     */
    getTableHeight() {
        return this.tt.height;
    }
    /**
     * @return {?}
     */
    getTableHeaderHeight() {
        return this.tt.headerBox.nativeElement.getBoundingClientRect().height;
    }
    /**
     * @return {?}
     */
    getTableBodyHeight() {
        return this.tt.height - this.getTableHeaderHeight();
    }
    /**
     * @return {?}
     */
    getTableWidth() {
        return this.tt.width;
    }
    /**
     * @return {?}
     */
    getRowHeight() {
        return this.tt.rowHeight;
    }
    /**
     * @param {?} scrollTop
     * @return {?}
     */
    getRowNodes(scrollTop) {
        /** @type {?} */
        let top = 0;
        /** @type {?} */
        const rows = [];
        /** @type {?} */
        let topHideHeight = 0;
        /** @type {?} */
        let bottomHideHeight = 0;
        /** @type {?} */
        const rowNodes = this.tt.getCurrentAllRowNodes();
        // const {rowNodes } = this.tt.state;
        /** @type {?} */
        const rowHeight = this.getRowHeight();
        /** @type {?} */
        const minTop = scrollTop;
        /** @type {?} */
        const maxTop = minTop + this.getTableHeight() + rowHeight;
        for (let i = 0; i < rowNodes.length; i++) {
            /** @type {?} */
            const n = rowNodes[i];
            if (!n.visible) {
                continue;
            }
            top += rowHeight;
            if (top + rowHeight < minTop) {
                topHideHeight += rowHeight;
                continue;
            }
            else {
                if (top > maxTop) {
                    bottomHideHeight += rowHeight;
                    continue;
                }
            }
            rows.push(n);
        }
        return {
            data: rows,
            topHideHeight,
            bottomHideHeight
        };
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    VirtualizedLoaderService.prototype.tt;
}
