// @farris/kendo-binding/esm5/lib/directives/farris_label_modification.directive.js
import { Directive, ElementRef, Injector, Input } from '@angular/core';
var FarrisLabelModificationDirective = /** @class */ (function () {
    function FarrisLabelModificationDirective(injector, elementRef) {
        this.injector = injector;
        this.elementRef = elementRef;
    }
    FarrisLabelModificationDirective.prototype.ngOnInit = function () {
        if (typeof this.index !== 'number' || !this.elementRef) {
            return;
        }
        var nativeElement = this.elementRef.nativeElement;
        if (nativeElement) {
            var labelFor = nativeElement.getAttribute('for');
            if (labelFor) {
                this.elementRef.nativeElement.setAttribute('for', labelFor + "_" + this.index);
            }
        }
    };
    FarrisLabelModificationDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-label-modification]'
                },] }
    ];
    /** @nocollapse */
    FarrisLabelModificationDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: ElementRef }
    ]; };
    FarrisLabelModificationDirective.propDecorators = {
        index: [{ type: Input }]
    };
    return FarrisLabelModificationDirective;
}());
export { FarrisLabelModificationDirective };
