// @farris/kendo-binding/esm5/lib/directives/farris-set-focus.directive.js
import { Directive, Input, Renderer2, Injector, ElementRef } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
var FarrisSetFocusDirective = /** @class */ (function () {
    function FarrisSetFocusDirective(rd, injector, el) {
        this.rd = rd;
        this.injector = injector;
        this.el = el;
        this._state = new BehaviorSubject({});
    }
    Object.defineProperty(FarrisSetFocusDirective.prototype, "focusState", {
        set: function (value) {
            this._state.next(value);
        },
        enumerable: true,
        configurable: true
    });
    FarrisSetFocusDirective.prototype.ngOnInit = function () {
        var _this = this;
        this.selfEl = this.el.nativeElement;
        this._state.subscribe(function (data) {
            if (data === false && _this.selfEl) {
                if (_this.selfEl.nodeName.toLowerCase() === 'farris-html-editor') {
                    _this.setHtmlEditorFocus();
                }
                else {
                    _this.setFocus();
                }
            }
        });
    };
    /**
     * 设置一般输入框的默认焦点
     */
    FarrisSetFocusDirective.prototype.setFocus = function () {
        var input = this.selfEl.querySelectorAll('input');
        if (input.length) {
            input[0].focus();
        }
        else {
            this.selfEl.focus();
        }
    };
    /**
     * 设置html编辑器的默认焦点  总是在文本最后
     */
    FarrisSetFocusDirective.prototype.setHtmlEditorFocus = function () {
        // 富文本编辑器不是input
        var ediotrArea = this.selfEl.querySelector('div[contenteditable="true"]');
        //创建range  存在浏览器兼容性问题
        var range = window.getSelection();
        //range 选择obj下所有子内容
        range.selectAllChildren(ediotrArea);
        //光标移至最后
        range.collapseToEnd();
    };
    FarrisSetFocusDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-set-focus]'
                },] }
    ];
    /** @nocollapse */
    FarrisSetFocusDirective.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: Injector },
        { type: ElementRef }
    ]; };
    FarrisSetFocusDirective.propDecorators = {
        focusState: [{ type: Input }]
    };
    return FarrisSetFocusDirective;
}());
export { FarrisSetFocusDirective };
