// @farris/kendo-binding/esm5/lib/directives/farris_lookup_binding.directive.js
import { Directive, Input, Optional } from '@angular/core';
import { ComboLookupComponent } from '@farris/ui-combo-lookup';
import { LookupGridComponent } from '@farris/ui-lookup';
var FarrisLookupBindingDirective = /** @class */ (function () {
    function FarrisLookupBindingDirective(lookup, comboLookup) {
        this.lookup = lookup;
        this.comboLookup = comboLookup;
    }
    Object.defineProperty(FarrisLookupBindingDirective.prototype, "lookupValue", {
        set: function (value) {
            if (this.lookupType === 'lookup' && this.lookup) {
                this.lookup.writeValue(value);
            }
            else if (this.lookupType === 'comboLookup' && this.comboLookup) {
                this.comboLookup.writeValue(value);
            }
        },
        enumerable: true,
        configurable: true
    });
    FarrisLookupBindingDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-lookup-binding]'
                },] }
    ];
    /** @nocollapse */
    FarrisLookupBindingDirective.ctorParameters = function () { return [
        { type: LookupGridComponent, decorators: [{ type: Optional }] },
        { type: ComboLookupComponent, decorators: [{ type: Optional }] }
    ]; };
    FarrisLookupBindingDirective.propDecorators = {
        lookupType: [{ type: Input }],
        lookupValue: [{ type: Input, args: ['farris-lookup-binding',] }]
    };
    return FarrisLookupBindingDirective;
}());
export { FarrisLookupBindingDirective };
