// @farris/kendo-binding/esm5/lib/directives/farris_checkbox_modification.directive.js
import { Directive, ElementRef, Injector, Input } from '@angular/core';
var FarrisCheckboxModificationDirective = /** @class */ (function () {
    function FarrisCheckboxModificationDirective(injector, elementRef) {
        this.injector = injector;
        this.elementRef = elementRef;
    }
    FarrisCheckboxModificationDirective.prototype.ngOnInit = function () {
        if (typeof this.index !== 'number' || !this.elementRef) {
            return;
        }
        var nativeElement = this.elementRef.nativeElement;
        if (nativeElement) {
            var id = nativeElement.getAttribute('id');
            if (id) {
                var token = id + "_" + this.index;
                this.elementRef.nativeElement.setAttribute('id', token);
                this.elementRef.nativeElement.setAttribute('name', token);
                this.elementRef.nativeElement.setAttribute('original_id', id);
            }
        }
    };
    FarrisCheckboxModificationDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-checkbox-modification]'
                },] }
    ];
    /** @nocollapse */
    FarrisCheckboxModificationDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: ElementRef }
    ]; };
    FarrisCheckboxModificationDirective.propDecorators = {
        index: [{ type: Input }]
    };
    return FarrisCheckboxModificationDirective;
}());
export { FarrisCheckboxModificationDirective };
