// @farris/kendo-binding/esm5/lib/directives/farris_editor.directive.js
import { Directive, ElementRef } from '@angular/core';
import { ViewModel } from '@farris/devkit';
import { EditorComponent } from '@farris/ui-editor';
var FarrisEditorDirective = /** @class */ (function () {
    function FarrisEditorDirective(viewModel, editor) {
        this.viewModel = viewModel;
        this.editor = editor;
    }
    FarrisEditorDirective.prototype.ngOnInit = function () {
        this.setComponentRef();
    };
    FarrisEditorDirective.prototype.setComponentRef = function () {
        var appContext = this.viewModel && this.viewModel.frameContext && this.viewModel.frameContext.getFormAppContext();
        var frameId = this.viewModel && this.viewModel.frameContext && this.viewModel.frameContext.frameId;
        var editor = this.editor;
        var elementRef = editor && editor.injector && editor.injector.get(ElementRef, null);
        var id = elementRef && elementRef.nativeElement && elementRef.nativeElement.id || null;
        // 如果frameId不存在或treetable没有id属性，说明不符合使用场景
        if (!frameId || !id) {
            return;
        }
        var map = appContext && appContext.componentRefs && appContext.componentRefs.get(frameId) || new Map();
        appContext && appContext.componentRefs && appContext.componentRefs.set(frameId, map.set(id, this.editor));
    };
    FarrisEditorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-editor],[farris-editor-bind]'
                },] }
    ];
    /** @nocollapse */
    FarrisEditorDirective.ctorParameters = function () { return [
        { type: ViewModel },
        { type: EditorComponent }
    ]; };
    return FarrisEditorDirective;
}());
export { FarrisEditorDirective };
