// @farris/kendo-binding/esm5/lib/directives/farris-sort.directive.js
import { Directive, Input } from '@angular/core';
var FarrisSortDirective = /** @class */ (function () {
    function FarrisSortDirective() {
    }
    FarrisSortDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farrisSort]'
                },] }
    ];
    return FarrisSortDirective;
}());
export { FarrisSortDirective };
var FarrisTextareaEditDirective = /** @class */ (function () {
    function FarrisTextareaEditDirective() {
    }
    FarrisTextareaEditDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-textarea-edit]'
                },] }
    ];
    return FarrisTextareaEditDirective;
}());
export { FarrisTextareaEditDirective };
var FarrisDynamicColumnFormatDirective = /** @class */ (function () {
    function FarrisDynamicColumnFormatDirective() {
    }
    FarrisDynamicColumnFormatDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farrisDynamicColumnFormat]'
                },] }
    ];
    return FarrisDynamicColumnFormatDirective;
}());
export { FarrisDynamicColumnFormatDirective };
var FarrisEditEnterDirective = /** @class */ (function () {
    function FarrisEditEnterDirective() {
    }
    FarrisEditEnterDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-enter-edit]'
                },] }
    ];
    FarrisEditEnterDirective.propDecorators = {
        farrisForm: [{ type: Input, args: ['form-group',] }]
    };
    return FarrisEditEnterDirective;
}());
export { FarrisEditEnterDirective };
var FarrisTreeTableFormatDirective = /** @class */ (function () {
    function FarrisTreeTableFormatDirective() {
    }
    FarrisTreeTableFormatDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farrisTreeTableFormat]'
                },] }
    ];
    return FarrisTreeTableFormatDirective;
}());
export { FarrisTreeTableFormatDirective };
