// @farris/kendo-binding/esm5/lib/directives/farris_querysolution_event_bind.directive.js
import { Directive, HostListener, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { QuerySolutionComponent } from "@gsp-cmp/querysolution";
var FarrisQuerySolutionEventBindDirective = /** @class */ (function () {
    function FarrisQuerySolutionEventBindDirective(injector, component, frameContext) {
        this.injector = injector;
        this.component = component;
        this.frameContext = frameContext;
    }
    /**
     * 执行筛选
     * @param event event
     */
    FarrisQuerySolutionEventBindDirective.prototype.onQuery = function (event) {
        if (this.frameContext) {
            this.frameContext.repository.entityCollection.pageIndex = 1;
        }
    };
    FarrisQuerySolutionEventBindDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[farris-querysolution-event-bind]"
                },] }
    ];
    /** @nocollapse */
    FarrisQuerySolutionEventBindDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: QuerySolutionComponent },
        { type: FrameContext }
    ]; };
    FarrisQuerySolutionEventBindDirective.propDecorators = {
        onQuery: [{ type: HostListener, args: ["queryEmitter",] }]
    };
    return FarrisQuerySolutionEventBindDirective;
}());
export { FarrisQuerySolutionEventBindDirective };
