// @farris/kendo-binding/esm5/lib/directives/farris-kendogrid-disabled.directive.js
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { GridComponent } from '@progress/kendo-angular-grid';
var FarrisDisabledDirective = /** @class */ (function () {
    function FarrisDisabledDirective(grid, el, renderer) {
        this.grid = grid;
        this.el = el;
        this.renderer = renderer;
    }
    FarrisDisabledDirective.prototype.ngOnInit = function () {
    };
    FarrisDisabledDirective.prototype.ngOnChanges = function (changes) {
        if (!this.maskElement) {
            this.createMaskElement();
        }
        if (changes['kendoGridDisabled'] && this.maskElement) {
            if (this.kendoGridDisabled) {
                this.renderer.removeStyle(this.maskElement, 'display');
            }
            else {
                this.renderer.setStyle(this.maskElement, 'display', 'none');
            }
        }
    };
    FarrisDisabledDirective.prototype.createMaskElement = function () {
        this.maskElement = this.renderer.createElement("div");
        this.renderer.setStyle(this.maskElement, 'width', '100%');
        this.renderer.setStyle(this.maskElement, 'height', '100%');
        this.renderer.setStyle(this.maskElement, 'position', 'absolute');
        this.renderer.setStyle(this.maskElement, 'background-color', 'rgba(255,255,255,0.5)');
        this.renderer.setStyle(this.maskElement, 'z-index', '3');
        this.renderer.setStyle(this.maskElement, 'display', 'none');
        this.renderer.appendChild(this.el.nativeElement, this.maskElement);
    };
    FarrisDisabledDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farrisDisabled]'
                },] }
    ];
    /** @nocollapse */
    FarrisDisabledDirective.ctorParameters = function () { return [
        { type: GridComponent },
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    FarrisDisabledDirective.propDecorators = {
        kendoGridDisabled: [{ type: Input, args: ['farrisDisabled',] }]
    };
    return FarrisDisabledDirective;
}());
export { FarrisDisabledDirective };
