// @farris/kendo-binding/esm5/lib/directives/grid-summary.directive.js
import { Directive, Input, NgZone } from '@angular/core';
import { BindingData } from '@farris/devkit';
import { ColumnComponent } from '@progress/kendo-angular-grid';
var GridColumnAggregateDirective = /** @class */ (function () {
    function GridColumnAggregateDirective(column, ngZone) {
        this.column = column;
        this.ngZone = ngZone;
    }
    Object.defineProperty(GridColumnAggregateDirective.prototype, "field", {
        get: function () {
            return this.column.field;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridColumnAggregateDirective.prototype, "aggrType", {
        get: function () {
            var type = 'none';
            if (typeof (this.aggregate) === 'boolean') {
                // 默认为sum，如果使用指令时传入的input为布尔类型，使用sum。
                if (this.aggregate) {
                    type = 'sum';
                }
            }
            else if (typeof (this.aggregate) === 'string') {
                type = this.aggregate;
            }
            return type;
        },
        enumerable: true,
        configurable: true
    });
    GridColumnAggregateDirective.prototype.recalculateSummary = function (bindingList) {
        var _this = this;
        var summary = null;
        if (this.aggrType === 'sum') {
            var plainData = bindingList.toJSON().map(function (item) {
                // 考虑帮助和udt的情况，field是a.b的格式。
                return _this.field.split('.').reduce(function (prevPath, currPath) {
                    return prevPath[currPath];
                }, item);
            });
            if (plainData && plainData.length) {
                summary = plainData.reduce(function (prev, curr) {
                    return prev + curr;
                });
            }
        }
        else if (this.aggrType === 'count') {
            // 计数，如果启用分页，只计算当前页，因此使用length而非total。
            summary = bindingList.length;
        }
        this.updateFooter(summary);
    };
    GridColumnAggregateDirective.prototype.updateFooter = function (content) {
        if (this.aggregate) {
            // 存在性能问题，去掉setTimeout。
            // 列表的列越多，此处问题越严重，需要正向解决以下错误。
            // todo: settimeout是为了解决数据变化时angular报错，如果更新页脚的逻辑不在gridbinding发起，是否还存在问题？
            // setTimeout(() => {
            this.column['summary'] = content;
            // });
        }
    };
    GridColumnAggregateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[aggregate]',
                },] }
    ];
    /** @nocollapse */
    GridColumnAggregateDirective.ctorParameters = function () { return [
        { type: ColumnComponent },
        { type: NgZone }
    ]; };
    GridColumnAggregateDirective.propDecorators = {
        aggregate: [{ type: Input }]
    };
    return GridColumnAggregateDirective;
}());
export { GridColumnAggregateDirective };
var GridSummaryDirective = /** @class */ (function () {
    function GridSummaryDirective(bindingData) {
        this.bindingData = bindingData;
    }
    GridSummaryDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farrisSummary]'
                },] }
    ];
    /** @nocollapse */
    GridSummaryDirective.ctorParameters = function () { return [
        { type: BindingData }
    ]; };
    GridSummaryDirective.propDecorators = {
        summary: [{ type: Input, args: ['farrisSummary',] }]
    };
    return GridSummaryDirective;
}());
export { GridSummaryDirective };
