// @gsp-wf/wf-process-editor/esm2015/lib/views/message-notice/message-notice.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ProcessDeUtil } from '../../domain/process-de-util';
export class MessageNotice {
    constructor() {
        /**
         * 接收人列表
         */
        this.receivers = new Array();
        this.id = ProcessDeUtil.GenerateElementId();
        this.subject = '消息通知';
        // this.sendType = 'email';
        this.content = `{"expr":"\'您有一条消息通知。\'","sexpr":""}`;
    }
}
if (false) {
    /** @type {?} */
    MessageNotice.prototype.id;
    /**
     * 消息内容
     * @type {?}
     */
    MessageNotice.prototype.content;
    /**
     * 消息发送时机
     * @type {?}
     */
    MessageNotice.prototype.sendTime;
    /**
     * 消息发送方式
     * @type {?}
     */
    MessageNotice.prototype.sendType;
    /**
     * 消息主题
     * @type {?}
     */
    MessageNotice.prototype.subject;
    /**
     * 消息接收人
     * @type {?}
     */
    MessageNotice.prototype.receiver;
    /**
     * 消息接收人显示
     * @type {?}
     */
    MessageNotice.prototype.receiverText;
    /**
     * 接收人列表
     * @type {?}
     */
    MessageNotice.prototype.receivers;
    /** @type {?} */
    MessageNotice.prototype.smsMessageTemplate;
}
