// @farris/mobile-command-services/esm5/lib/exception-services/http.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/exception-services/http.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var HttpStatusType = {
    Info: 1,
    Success: 2,
    Redirect: 3,
    ClientError: 4,
    ServerError: 5,
};
HttpStatusType[HttpStatusType.Info] = 'Info';
HttpStatusType[HttpStatusType.Success] = 'Success';
HttpStatusType[HttpStatusType.Redirect] = 'Redirect';
HttpStatusType[HttpStatusType.ClientError] = 'ClientError';
HttpStatusType[HttpStatusType.ServerError] = 'ServerError';
/**
 * http通用类
 */
var /**
 * http通用类
 */
HttpUtil = /** @class */ (function () {
    function HttpUtil() {
    }
    /**
     * 基于http状态码返回状态类型
     * @param httpResponseSatus http请求返回状态码
     * @returns null | HttpStatusType
     */
    /**
     * 基于http状态码返回状态类型
     * @param {?} httpResponseSatus http请求返回状态码
     * @return {?} null | HttpStatusType
     */
    HttpUtil.getHttpStatusType = /**
     * 基于http状态码返回状态类型
     * @param {?} httpResponseSatus http请求返回状态码
     * @return {?} null | HttpStatusType
     */
    function (httpResponseSatus) {
        if (!httpResponseSatus) {
            return null;
        }
        /** @type {?} */
        var statusType = null;
        if (httpResponseSatus < 100 || httpResponseSatus > 600) {
            statusType = null; // 不合理状态码
        }
        else if (httpResponseSatus >= 100 && httpResponseSatus < 200) {
            statusType = HttpStatusType.Info;
        }
        else if (httpResponseSatus >= 200 && httpResponseSatus < 300) {
            statusType = HttpStatusType.Success;
        }
        else if (httpResponseSatus >= 300 && httpResponseSatus < 400) {
            statusType = HttpStatusType.Redirect;
        }
        else if (httpResponseSatus >= 400 && httpResponseSatus < 500) {
            statusType = HttpStatusType.ClientError;
        }
        else if (httpResponseSatus >= 500 && httpResponseSatus < 600) {
            statusType = HttpStatusType.ServerError;
        }
        return statusType;
    };
    return HttpUtil;
}());
export { HttpUtil, HttpStatusType };
