// @farris/ui-loading/esm5/lib/loading.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/**
 * @record
 */
export function LoadingConfig() { }
if (false) {
    /** @type {?|undefined} */
    LoadingConfig.prototype.width;
    /** @type {?|undefined} */
    LoadingConfig.prototype.container;
    /** @type {?|undefined} */
    LoadingConfig.prototype.message;
    /** @type {?|undefined} */
    LoadingConfig.prototype.position;
    /** @type {?|undefined} */
    LoadingConfig.prototype.template;
    /** @type {?|undefined} */
    LoadingConfig.prototype.type;
    /**
     * 延迟1s 后在显示
     * @type {?|undefined}
     */
    LoadingConfig.prototype.delay;
}
/** @type {?} */
export var loaddingDefaultConfig = {
    width: 30,
    container: 'body',
    message: '正在加载中，请稍候...',
    posion: 'center',
    type: 0,
    delay: 300
};
/** @type {?} */
export var LOADING_DEFAULT_CONFIG = new InjectionToken('loading default options.');
