// @farris/ui-loading/esm5/lib/loading.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LoadingComponent } from './loading.component';
import { LoadingService } from './loading.service';
import { LOADING_DEFAULT_CONFIG, loaddingDefaultConfig } from './loading.config';
import { FarrisCommonModule } from '@farris/ui-common';
import { LocaleModule } from '@farris/ui-locale';
var LoadingModule = /** @class */ (function () {
    function LoadingModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    LoadingModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: LoadingModule,
            providers: [
                { provide: LOADING_DEFAULT_CONFIG, useValue: config || loaddingDefaultConfig },
                LoadingService
            ]
        };
    };
    LoadingModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FarrisCommonModule,
                        LocaleModule
                    ],
                    declarations: [
                        LoadingComponent
                    ],
                    entryComponents: [
                        LoadingComponent
                    ]
                },] }
    ];
    return LoadingModule;
}());
export { LoadingModule };
