// @farris/ui-filter-condition/esm5/lib/handlers/text.handler.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CompareType } from '../enum/compare-type.enum';
import { ValueType } from '../enum/value-type.enum';
import { RelationType } from '../enum/relation-type.enum';
var TextHandler = /** @class */ (function () {
    function TextHandler() {
    }
    /**
     * @param {?} qc
     * @return {?}
     */
    TextHandler.prototype.convert = /**
     * @param {?} qc
     * @return {?}
     */
    function (qc) {
        /** @type {?} */
        var tv = (qc.value);
        return Object.keys(tv).length > 0 && tv.value ? [{
                FilterField: qc.labelCode,
                Compare: CompareType.Like,
                // like CompareType.like undefined
                Value: tv.value,
                Relation: RelationType.And,
                // and undefined
                Expresstype: ValueType.Value
            }] : [];
    };
    return TextHandler;
}());
export { TextHandler };
