// @gsp-sys/rtf-apphelp/esm2015/lib/common/rtf-ui-webapi.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class RtfUIWebapiService {
    constructor() { }
    /**
     * 菜单api地址
     * @return {?}
     */
    get functionsApi() {
        return '/api/runtime/sys/v1.0/gspapp/all';
    }
    /**
     * 带有BO数据的应用 访问地址
     * @return {?}
     */
    get appsWithBOApi() {
        return '/api/runtime/sys/v1.0/gspapp/allAppsBO';
    }
    /**
     * 业务对象api地址
     * @return {?}
     */
    get bizObjectApi() {
        return '/api/dev/main/v1.0/business-objects';
    }
    /**
     * Http头信息
     * @return {?}
     */
    get httpHeader() {
        /** @type {?} */
        const options = {
            headers: {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                SessionId: this.sessionId
            },
            responseType: 'json'
        };
        return options;
    }
}
RtfUIWebapiService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RtfUIWebapiService.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    RtfUIWebapiService.prototype.sessionId;
}
