// @edp-bif/common-api/esm2015/lib/entity/TaskCenterConfig.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/TaskCenterConfig.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl, TaskCenterContent, TaskHandlingMenuArgs } from "@edp-aif/common-api";
import { JointQueryMode } from "./impl/JointQueryMode";
export class TaskCenterConfig extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.jointQueryMode = JointQueryMode.APP;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.taskCenterContent = null;
        this.jointQueryMode = JointQueryMode.APP;
        this.joitQueryMenu = null;
        this.jointQueryUrl = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterContent", this.taskCenterContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryMode", this.jointQueryMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "joitQueryMenu", this.joitQueryMenu);
        BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryUrl", this.jointQueryUrl);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.taskCenterContent = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "taskCenterContent", TaskCenterContent);
        this.jointQueryMode = JointQueryMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryMode")];
        this.joitQueryMenu = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "joitQueryMenu", TaskHandlingMenuArgs);
        this.jointQueryUrl = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryUrl");
    }
}
if (false) {
    /** @type {?} */
    TaskCenterConfig.prototype.taskCenterContent;
    /** @type {?} */
    TaskCenterConfig.prototype.jointQueryMode;
    /** @type {?} */
    TaskCenterConfig.prototype.jointQueryUrl;
    /** @type {?} */
    TaskCenterConfig.prototype.joitQueryMenu;
}
