// @gsp-cmp/coderule-selector/esm5/lib/components/code-rule-list/code-rule-list.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/code-rule-list/code-rule-list.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { process } from '@progress/kendo-data-query';
import { CodeRuleBaseInfo } from '../../model/code-rule-base-info';
import { CodeRuleSelectService } from '../../service/code-rule-select.service';
var CodeRuleListComponent = /** @class */ (function () {
    function CodeRuleListComponent(codeRuleSvc) {
        this.codeRuleSvc = codeRuleSvc;
        this.data = new Array();
        // tslint:disable-next-line:member-ordering
        this.state = {
            skip: 0,
            take: 9999,
            // Initial filter descriptor
            filter: {
                logic: 'and',
                filters: []
            }
        };
        // tslint:disable-next-line:member-ordering
        this.selectableSetting = { mode: 'single', checkboxOnly: false };
    }
    /**
     * @return {?}
     */
    CodeRuleListComponent.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.data = [];
    };
    /**
     * @param {?} state
     * @return {?}
     */
    CodeRuleListComponent.prototype.dataStateChange = /**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this.state = state;
        this.gridData = process(this.data, this.state);
    };
    /**
     * @return {?}
     */
    CodeRuleListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.codeRuleSvc.getCodeRuleList().subscribe((/**
         * @param {?} datas
         * @return {?}
         */
        function (datas) {
            if (datas && datas.length > 0) {
                datas.forEach((/**
                 * @param {?} obj
                 * @return {?}
                 */
                function (obj) {
                    /** @type {?} */
                    var baseInfo = new CodeRuleBaseInfo();
                    baseInfo.init(obj);
                    _this.data.push(baseInfo);
                }));
                _this.gridData = process(_this.data, _this.state);
            }
        }));
    };
    // 选中编号规则后的事件处理
    // 选中编号规则后的事件处理
    /**
     * @param {?} event
     * @return {?}
     */
    CodeRuleListComponent.prototype.onSelectionChange = 
    // 选中编号规则后的事件处理
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.selectedRows.length > 0) {
            this.codeRuleSvc.codeRule = null;
            this.codeRuleSvc.codeRule = event.selectedRows[0].dataItem;
        }
    };
    CodeRuleListComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'app-coderule-list',
                    template: "<div class=\"templateInfo\">\r\n  <div class=\"templateSelector\">\r\n    <kendo-grid [data]=\"gridData\" [height]=\"449\" [selectable]=\"selectableSetting\"\r\n      (selectionChange)=onSelectionChange($event) [pageSize]=\"state.take\" [skip]=\"state.skip\" [sort]=\"state.sort\"\r\n      [filter]=\"state.filter\" [sortable]=\"true\" [pageable]=\"false\" [filterable]=\"true\" [resizable]=\"true\"\r\n      (dataStateChange)=dataStateChange($event)>\r\n      <kendo-grid-column field=\"code\" title=\"\u7F16\u53F7\" width=\"210\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column field=\"name\" title=\"\u540D\u79F0\" width=\"200\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column field=\"serviceUnit\" title=\"\u670D\u52A1\u5355\u5143\" width=\"200\">\r\n      </kendo-grid-column>\r\n    </kendo-grid>\r\n  </div>\r\n</div>",
                    styles: [".templateSelector{width:100%;height:450px;border:.5px solid #ddd}.templateDescription{width:200px;height:450px;border:.5px solid #ddd}.templateInfo{display:flex}"]
                }] }
    ];
    /** @nocollapse */
    CodeRuleListComponent.ctorParameters = function () { return [
        { type: CodeRuleSelectService }
    ]; };
    return CodeRuleListComponent;
}());
export { CodeRuleListComponent };
if (false) {
    /** @type {?} */
    CodeRuleListComponent.prototype.data;
    /** @type {?} */
    CodeRuleListComponent.prototype.state;
    /** @type {?} */
    CodeRuleListComponent.prototype.gridData;
    /** @type {?} */
    CodeRuleListComponent.prototype.selectableSetting;
    /**
     * @type {?}
     * @private
     */
    CodeRuleListComponent.prototype.codeRuleSvc;
}
