// @farris/ui-list-filter/esm5/lib/entity/list-filter.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var listFilterItem = /** @class */ (function () {
    function listFilterItem(_id, _groupName) {
        this.id = _id;
        this.name = _groupName;
        this.width = 0;
        this.isExtend = false;
    }
    // 设置宽度
    // 设置宽度
    /**
     * @param {?} value
     * @return {?}
     */
    listFilterItem.prototype.setWidth = 
    // 设置宽度
    /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.width = parseInt(value, 10);
    };
    // 获取宽度
    // 获取宽度
    /**
     * @return {?}
     */
    listFilterItem.prototype.getWidth = 
    // 获取宽度
    /**
     * @return {?}
     */
    function () {
        return this.width;
    };
    return listFilterItem;
}());
export { listFilterItem };
if (false) {
    /** @type {?} */
    listFilterItem.prototype.id;
    /** @type {?} */
    listFilterItem.prototype.name;
    /** @type {?} */
    listFilterItem.prototype.width;
    /** @type {?} */
    listFilterItem.prototype.isExtend;
}
var listFilterGroup = /** @class */ (function () {
    function listFilterGroup() {
        this.responsedIndex = 0;
        this.listItems = [];
    }
    return listFilterGroup;
}());
export { listFilterGroup };
if (false) {
    /** @type {?} */
    listFilterGroup.prototype.listItems;
    /** @type {?} */
    listFilterGroup.prototype.responsedIndex;
}
