// @farris/ui-time-picker/esm5/lib/time-value-accessor.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { parseTime, format } from './utils/date-helper';
var TimeValueAccessorDirective = /** @class */ (function () {
    function TimeValueAccessorDirective(elementRef) {
        this.elementRef = elementRef;
    }
    /**
     * @return {?}
     */
    TimeValueAccessorDirective.prototype.keyup = /**
     * @return {?}
     */
    function () {
        this.changed();
    };
    /**
     * @return {?}
     */
    TimeValueAccessorDirective.prototype.blur = /**
     * @return {?}
     */
    function () {
        this.touched();
    };
    /**
     * @return {?}
     */
    TimeValueAccessorDirective.prototype.changed = /**
     * @return {?}
     */
    function () {
        if (this._onChange) {
            /** @type {?} */
            var value = parseTime(this.elementRef.nativeElement.value);
            this._onChange((/** @type {?} */ (value)));
        }
    };
    /**
     * @return {?}
     */
    TimeValueAccessorDirective.prototype.touched = /**
     * @return {?}
     */
    function () {
        if (this._onTouch) {
            this._onTouch();
        }
    };
    /**
     * @return {?}
     */
    TimeValueAccessorDirective.prototype.setRange = /**
     * @return {?}
     */
    function () {
        this.elementRef.nativeElement.focus();
        this.elementRef.nativeElement.setSelectionRange(0, this.elementRef.nativeElement.value.length);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TimeValueAccessorDirective.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.elementRef.nativeElement.value = format(value, this.nzTime);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TimeValueAccessorDirective.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TimeValueAccessorDirective.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._onTouch = fn;
    };
    TimeValueAccessorDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'input[format]',
                    exportAs: 'format',
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: TimeValueAccessorDirective,
                            multi: true
                        }
                    ]
                },] }
    ];
    /** @nocollapse */
    TimeValueAccessorDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    TimeValueAccessorDirective.propDecorators = {
        nzTime: [{ type: Input }],
        keyup: [{ type: HostListener, args: ['keyup',] }],
        blur: [{ type: HostListener, args: ['blur',] }]
    };
    return TimeValueAccessorDirective;
}());
export { TimeValueAccessorDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TimeValueAccessorDirective.prototype._onChange;
    /**
     * @type {?}
     * @private
     */
    TimeValueAccessorDirective.prototype._onTouch;
    /** @type {?} */
    TimeValueAccessorDirective.prototype.nzTime;
    /**
     * @type {?}
     * @private
     */
    TimeValueAccessorDirective.prototype.elementRef;
}
