// @farris/mobile-vue-adapter/esm5/lib/attachment/attachment.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/attachment/attachment.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
/**
 * @abstract
 */
var /**
 * @abstract
 */
AttachmentService = /** @class */ (function () {
    function AttachmentService() {
        this.collectFile = new Subject();
    }
    /**
     * 拍照
     * @param param 拍照参数
     */
    /**
     * 拍照
     * @param {?=} param 拍照参数
     * @return {?}
     */
    AttachmentService.prototype.takePhoto = /**
     * 拍照
     * @param {?=} param 拍照参数
     * @return {?}
     */
    function (param) {
        this.takePhotoIns.takePhotoAttachment(param);
    };
    /**
     * 选择文件
     * @param param 选择文件参数
     */
    /**
     * 选择文件
     * @param {?} param 选择文件参数
     * @return {?}
     */
    AttachmentService.prototype.selectAttachment = /**
     * 选择文件
     * @param {?} param 选择文件参数
     * @return {?}
     */
    function (param) {
        this.takePhotoIns.selectAttachments(param);
    };
    /**
     * 拍摄短视频
     * @param param 拍摄短视频参数
     */
    /**
     * 拍摄短视频
     * @param {?} param 拍摄短视频参数
     * @return {?}
     */
    AttachmentService.prototype.takeShortVideo = /**
     * 拍摄短视频
     * @param {?} param 拍摄短视频参数
     * @return {?}
     */
    function (param) {
        this.takePhotoIns.takeShortVideo(param);
    };
    /**
     * 播放短视频
     * @param param 播放短视频参数
     */
    /**
     * 播放短视频
     * @param {?} param 播放短视频参数
     * @return {?}
     */
    AttachmentService.prototype.playShortVideo = /**
     * 播放短视频
     * @param {?} param 播放短视频参数
     * @return {?}
     */
    function (param) {
        this.takePhotoIns.playShortVideo(param);
    };
    /**
     * @param {?} param
     * @return {?}
     */
    AttachmentService.prototype.selectFile = /**
     * @param {?} param
     * @return {?}
     */
    function (param) {
        this.takePhotoIns.selectFile(param);
    };
    /**
     * @param {?} param
     * @return {?}
     */
    AttachmentService.prototype.downloadAttachment = /**
     * @param {?} param
     * @return {?}
     */
    function (param) {
        this.takePhotoIns.downloadAttachment(param);
    };
    /**
     * @return {?}
     */
    AttachmentService.prototype.getSysInfo = /**
     * @return {?}
     */
    function () {
        return this.takePhotoIns.getSysInfo();
    };
    return AttachmentService;
}());
/**
 * @abstract
 */
export { AttachmentService };
if (false) {
    /** @type {?} */
    AttachmentService.prototype.takePhotoIns;
    /** @type {?} */
    AttachmentService.prototype.collectFile;
}
