// @farris/mobile-vue-adapter/esm5/lib/attachment/attachment-filter.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/attachment/attachment-filter.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { getDataByPath, getAttachmentUrl } from '../utils';
/** @type {?} */
export var filterData = (/**
 * @param {?} rawData
 * @param {?} field
 * @param {?} rootId
 * @param {?=} isComponentAdapter
 * @return {?}
 */
function (rawData, field, rootId, isComponentAdapter) {
    if (isComponentAdapter === void 0) { isComponentAdapter = true; }
    /** @type {?} */
    var splitFiled = field.split('/');
    /** @type {?} */
    var fields = splitFiled.filter((/**
     * @param {?} ele
     * @return {?}
     */
    function (ele) { return ele; }));
    if (!isComponentAdapter) {
        // 单条数据
        /** @type {?} */
        var data = getDataByPath(rawData, fields);
        /** @type {?} */
        var url = '/platform/common/web/assets/imgs/no-pic.png';
        // 附件不存在
        if (!data || !data.attachmentId) {
            return url;
        }
        // 附件存在
        /** @type {?} */
        var extensionName = data.fileName.substring(data.fileName.lastIndexOf('.') + 1);
        url = getAttachmentUrl(data.attachmentId, rootId, extensionName);
        return url;
    }
    /** @type {?} */
    var transformData = rawData.map((/**
     * @param {?} everyData
     * @return {?}
     */
    function (everyData) {
        /** @type {?} */
        var data = getDataByPath(everyData, fields);
        /** @type {?} */
        var url = '/platform/common/web/assets/imgs/no-pic.png';
        if (!data || !data.attachmentId) {
            return {
                url: url,
                id: everyData.id,
                attachmentId: data.attachmentId,
                name: '',
                size: 0,
                createTime: ''
            };
        }
        /** @type {?} */
        var extensionName = data.fileName.substring(data.fileName.lastIndexOf('.') + 1);
        url = getAttachmentUrl(data.attachmentId, rootId, extensionName);
        return {
            url: url,
            id: everyData.id,
            attachmentId: data.attachmentId,
            name: data.fileName,
            size: data.fileSize,
            createTime: data.fileCreateTime
        };
    }));
    return transformData;
});
