// @farris/mobile-devkit/esm2015/lib/form/form_control.js
import { BindingValueAccessorFactory } from './binding_value_accessor';
import { ValidatorFactory } from '../validator/index';
/**
 * FormControl定义
 */
class FormControl {
    constructor(config, viewModelContext) {
        this.validatorFn = [];
        this.required = false;
        this.requiredExpressionFlag = false;
        this.validateExpressionFlag = false;
        this.readonly = false;
        this.visible = true;
        // 初始验证规则
        this.validRules = [];
        // 记忆验证表达式的规则
        this.validRules_validates = [];
        this.valueAccessor = BindingValueAccessorFactory.create(config.bindingType, config.bindingPath, config.valueConverter, viewModelContext);
        this.validRules = [];
        if (config.validRules) {
            this.validRules = config.validRules;
            this.setValidatorFn(config.validRules);
        }
    }
    get value() {
        return this.valueAccessor.getValue();
    }
    set value(val) {
        this.valueAccessor.setValue(val);
    }
    getValidRules() {
        return this.validRules;
    }
    setValidatorFn(validRules) {
        this.validatorFn = ValidatorFactory.create(validRules);
    }
    pushValidatorFnforValidate(validRules, flag) {
        if (flag && this.validateExpressionFlag) {
            return;
        }
        this.validateExpressionFlag = flag;
        this.validRules_validates = validRules;
        this.validatorFn = this.validatorFn.concat(ValidatorFactory.create(validRules));
    }
    pushValidatorFnforRequired(validRules, flag) {
        if (flag && this.requiredExpressionFlag) {
            return;
        }
        this.requiredExpressionFlag = flag;
        this.validatorFn = this.validatorFn.concat(ValidatorFactory.create(validRules));
    }
    resetValidatorFnforRequired() {
        this.requiredExpressionFlag = false;
        this.validatorFn = ValidatorFactory.create(this.validRules.concat(this.validRules_validates));
    }
    getValidatorFn() {
        return this.validatorFn;
    }
}
export { FormControl };
