// @gsp-cmp/common-component/esm5/convertor/util/ref-structure-util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { RefStructureInfo } from '../../define/ref-structure-info';
var RefStructureUtil = /** @class */ (function () {
    function RefStructureUtil() {
    }
    /**
     * @param {?} refInfo
     * @return {?}
     */
    RefStructureUtil.convertToObject = /**
     * @param {?} refInfo
     * @return {?}
     */
    function (refInfo) {
        /** @type {?} */
        var result = new Object();
        result["id"] = refInfo.id;
        result["referId"] = refInfo.referId;
        result["referType"] = refInfo.referType;
        result["referCode"] = refInfo.referCode;
        result["structuredTypeId"] = refInfo.structuredTypeId;
        result["code"] = refInfo.code;
        result["name"] = refInfo.name;
        return result;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RefStructureUtil.initFromObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var result = new RefStructureInfo();
        result.id = obj['id'];
        result.referId = obj['referId'];
        result.referType = obj['referType'];
        result.referCode = obj['referCode'];
        result.structuredTypeId = obj['structuredTypeId'];
        result.code = obj['code'];
        result.name = obj['name'];
        return result;
    };
    return RefStructureUtil;
}());
export { RefStructureUtil };
