// @gsp-cmp/coderule-selector/esm5/lib/components/select-code-rule/select-code-rule.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/select-code-rule/select-code-rule.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// tslint:disable-next-line:max-line-length
import { Component, Input, Output, EventEmitter, ComponentFactoryResolver, Injector } from '@angular/core';
import { CodeRuleSelectService } from '../../service/code-rule-select.service';
import { CodeRuleInvokeInfo } from '../../model/code-rule-invoke-info';
var SelectCodeRuleComponent = /** @class */ (function () {
    function SelectCodeRuleComponent(codeRuleSvc, resolver, injector) {
        this.codeRuleSvc = codeRuleSvc;
        this.resolver = resolver;
        this.injector = injector;
        this.codeRuleSelected = new EventEmitter();
        this.selectorClosed = new EventEmitter();
    }
    Object.defineProperty(SelectCodeRuleComponent.prototype, "serviceUnit", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.codeRuleSvc.serviceUnit = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SelectCodeRuleComponent.prototype, "sessionId", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.codeRuleSvc.sessionId = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} changes
     * @return {?}
     */
    SelectCodeRuleComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
    };
    /**
     * 点击确认，确认编号规则
     */
    /**
     * 点击确认，确认编号规则
     * @return {?}
     */
    SelectCodeRuleComponent.prototype.confirmCodeRule = /**
     * 点击确认，确认编号规则
     * @return {?}
     */
    function () {
        /** @type {?} */
        var codeRuleInfo = this.codeRuleSvc.codeRule;
        if (!codeRuleInfo) {
            alert("当前未选中任何数据！");
            return;
        }
        /** @type {?} */
        var invokeInfo = new CodeRuleInvokeInfo();
        invokeInfo.id = codeRuleInfo.id;
        invokeInfo.code = codeRuleInfo.code;
        invokeInfo.name = codeRuleInfo.name;
        invokeInfo.serviceUnit = codeRuleInfo.serviceUnit;
        invokeInfo.byMetadata = codeRuleInfo.origin == "元数据" ? true : false;
        this.codeRuleSelected.emit(invokeInfo);
        this.closeSelector();
        // 清空缓存
        this.codeRuleSvc.clearData();
    };
    /**
     * 点击取消，关闭选择器
     */
    /**
     * 点击取消，关闭选择器
     * @return {?}
     */
    SelectCodeRuleComponent.prototype.closeSelector = /**
     * 点击取消，关闭选择器
     * @return {?}
     */
    function () {
        this.selectorClosed.emit();
    };
    SelectCodeRuleComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'app-select-metadatart',
                    template: "<div class=\"modal-body\">\r\n    <div class=\"templateBlock\">\r\n        <app-coderule-list ></app-coderule-list>\r\n    </div>\r\n</div>\r\n<div class=\"modal-footer\">\r\n    <button type=\"button\" class=\"btn btn-default btn-primary\" (click)=\"confirmCodeRule()\">\u786E\u5B9A</button>\r\n    <button type=\"button\" class=\"btn btn-default\" (click)=\"closeSelector()\">\u53D6\u6D88</button>\r\n</div>",
                    styles: [".templateBlock{display:flex}.infoBlock{width:100%;min-width:400px}.infoBlock label{width:50px}"]
                }] }
    ];
    /** @nocollapse */
    SelectCodeRuleComponent.ctorParameters = function () { return [
        { type: CodeRuleSelectService },
        { type: ComponentFactoryResolver },
        { type: Injector }
    ]; };
    SelectCodeRuleComponent.propDecorators = {
        serviceUnit: [{ type: Input }],
        sessionId: [{ type: Input }],
        selectorClosed: [{ type: Output }],
        codeRuleSelected: [{ type: Output }]
    };
    return SelectCodeRuleComponent;
}());
export { SelectCodeRuleComponent };
if (false) {
    /** @type {?} */
    SelectCodeRuleComponent.prototype.selectorClosed;
    /** @type {?} */
    SelectCodeRuleComponent.prototype.codeRuleSelected;
    /**
     * @type {?}
     * @private
     */
    SelectCodeRuleComponent.prototype.codeRuleSvc;
    /**
     * @type {?}
     * @private
     */
    SelectCodeRuleComponent.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    SelectCodeRuleComponent.prototype.injector;
}
