// @farris/mobile-devkit/esm2015/lib/expression/expression/expression_result_factory.js
import { ExpressionManager } from "./expression_manager";
;
import { ExpressionRegistry } from "./expression_registry";
import { ExpressionResult } from "./expression_result";
export class ExpressionResultFactory {
    constructor(viewModelContext) {
        this.viewModelContext = viewModelContext;
        this.injector = this.viewModelContext.injector;
        this.expressionRegistry = this.injector.get(ExpressionRegistry);
        this.expressionManager = this.injector.get(ExpressionManager);
        this.expressionResult = this.injector.get(ExpressionResult);
        this.registeEvent();
    }
    registeEvent() {
        this.expressionRegistry.expressions.subscribe((expressionObjects) => {
            // 加载完表达式之后做一次计算
            expressionObjects.forEach((expressionObject) => {
                if (expressionObject.deps && expressionObject.deps.length > 0) {
                    return;
                }
                const result = this.expressionManager.eval(expressionObject.id);
                this.expressionResult[expressionObject.id] = result;
            });
        });
    }
}
