// @farris/mobile-devkit/esm2015/lib/expression/expression/expression_event_emitter.js
import { BehaviorSubject } from "rxjs";
import { Listeners } from "../listener/index";
/**
 * 事件监听器主要有以下几种类型的事件：
 * 1、值变化
 * 2、行切换
 *    行切换使用场景为需要为数据计算依赖当前行时
 * 3、数据加载
 */
export class ExpressionEventEmitter {
    constructor(viewModelContext) {
        this.viewModelContext = viewModelContext;
        this.listeners = this.viewModelContext.injector.get(Listeners);
        this.events = new Array();
        this.listeners.onEvent.subscribe((eventArgs) => {
            if (this.onEvent && this.onEvent.observers.length > 0) {
                let events = [];
                if (this.events.length > 0) {
                    events = [...this.events];
                }
                events.push(eventArgs);
                this.onEvent.next(events);
                this.events = [];
            }
            else {
                this.events.push(eventArgs);
            }
        });
    }
    attach() {
        if (!this.onEvent) {
            this.onEvent = new BehaviorSubject(this.events);
        }
        return this.onEvent.asObservable();
    }
}
