// @farris/ide-devkit/esm2015/lib/components/ng2tree-common/components/tree-node-content/tree-node-content.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ComponentFactoryResolver, ViewContainerRef, ViewChild } from '@angular/core';
export class TreeNodeContentComponent {
    /**
     * @param {?} componentFactoryResolver
     */
    constructor(componentFactoryResolver) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.rightMenuClicked = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // 右键菜单相关
        this.rightMenuContainer.clear();
        if (this.originData.hasRightClickMenu) {
            /** @type {?} */
            const componentFactory = this.componentFactoryResolver.resolveComponentFactory(this.originData.rightClickMenuComponent);
            /** @type {?} */
            const componentRef = this.rightMenuContainer.createComponent(componentFactory);
            componentRef.instance['nodeData'] = this.originData;
            componentRef.instance['rightMenuClicked'].subscribe((event) => {
                this.rightMenuClicked.emit(event);
            });
        }
    }
}
TreeNodeContentComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng2tree-node-content',
                template: "<div class='tree-node-content' *ngIf=\"!originData.hasRightClickMenu\">\r\n  <span>{{ displayData }}</span>\r\n</div>\r\n<ng-container #rightMenuContainer>\r\n</ng-container>",
                styles: [".tree-node-content:hover{background:#a8daa7}"]
            }] }
];
/** @nocollapse */
TreeNodeContentComponent.ctorParameters = () => [
    { type: ComponentFactoryResolver }
];
TreeNodeContentComponent.propDecorators = {
    displayData: [{ type: Input }],
    originData: [{ type: Input }],
    rightMenuClicked: [{ type: Output }],
    rightMenuContainer: [{ type: ViewChild, args: ['rightMenuContainer', { read: ViewContainerRef },] }]
};
if (false) {
    /** @type {?} */
    TreeNodeContentComponent.prototype.displayData;
    /** @type {?} */
    TreeNodeContentComponent.prototype.originData;
    /** @type {?} */
    TreeNodeContentComponent.prototype.rightMenuClicked;
    /** @type {?} */
    TreeNodeContentComponent.prototype.rightMenuContainer;
    /**
     * @type {?}
     * @private
     */
    TreeNodeContentComponent.prototype.componentFactoryResolver;
}
