// @ecp-caf/common-structure/esm5/lib/convertor/structure/common-structure-convertor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { StructuredTypeConvertor } from '../data-type/structured-type-convertor';
import { ReferReplaceUtil } from '../utils/refer-replace-util';
var CommonStructureConvertor = /** @class */ (function () {
    function CommonStructureConvertor() {
    }
    /**
     * @param {?} obj
     * @param {?} structure
     * @param {?} ctxt
     * @return {?}
     */
    CommonStructureConvertor.prototype.convertToObject = /**
     * @param {?} obj
     * @param {?} structure
     * @param {?} ctxt
     * @return {?}
     */
    function (obj, structure, ctxt) {
        if (!structure) {
            return;
        }
        obj['id'] = structure.id;
        obj['uri'] = structure.uri;
        obj['code'] = structure.code;
        obj['name'] = structure.name;
        obj['kind'] = structure.kind;
        obj['description'] = structure.description;
        /** @type {?} */
        var structureConvertor = new StructuredTypeConvertor();
        obj['structuredTypes'] = new Array();
        if (structure.structuredTypes) {
            structure.structuredTypes.forEach((/**
             * @param {?} structuredType
             * @return {?}
             */
            function (structuredType) {
                /** @type {?} */
                var kind = structuredType.kind;
                /** @type {?} */
                var serializer = ctxt.getDataTypeSerializerFactory().get(kind);
                /** @type {?} */
                var structuredTypeObj = serializer.serialize(structuredType, ctxt);
                obj['structuredTypes'].push(structuredTypeObj);
            }));
        }
        obj['refStructures'] = new Array();
        if (structure.refStructures) {
            structure.refStructures.forEach((/**
             * @param {?} refStructure
             * @return {?}
             */
            function (refStructure) {
                /** @type {?} */
                var kind = refStructure.kind;
                /** @type {?} */
                var serializer = ctxt.getStructureSerializerFactory().get(kind);
                /** @type {?} */
                var refStructureObj = serializer.serialize(refStructure, ctxt);
                obj['refStructures'].push(refStructureObj);
            }));
        }
    };
    /**
     * @param {?} structure
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    CommonStructureConvertor.prototype.initFromObject = /**
     * @param {?} structure
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    function (structure, obj, ctxt) {
        if (!obj) {
            return;
        }
        structure.kind = obj['kind'];
        structure.id = obj['id'];
        structure.uri = obj['uri'];
        structure.code = obj['code'];
        structure.name = obj['name'];
        structure.description = obj['description'];
        if (obj['structuredTypes']) {
            structure.structuredTypes = new Array();
            obj['structuredTypes'].forEach((/**
             * @param {?} structureObj
             * @return {?}
             */
            function (structureObj) {
                /** @type {?} */
                var kind = structureObj['kind'];
                /** @type {?} */
                var serializer = ctxt.getDataTypeSerializerFactory().get(kind);
                /** @type {?} */
                var structuredType = (/** @type {?} */ (serializer.deserialize(structureObj, ctxt)));
                structure.structuredTypes.push(structuredType);
                // structure.structuredTypes.push(DataTypeConvertUtil.initFromObject(structureObj) as StructuredType);
            }));
        }
        if (obj['refStructures']) {
            structure.refStructures = new Array();
            obj['refStructures'].forEach((/**
             * @param {?} refTree
             * @return {?}
             */
            function (refTree) {
                /** @type {?} */
                var kind = refTree['kind'];
                /** @type {?} */
                var refSerializer = ctxt.getStructureSerializerFactory().get(kind);
                /** @type {?} */
                var refStructure = refSerializer.deserialize(refTree, ctxt);
                structure.refStructures.push(refStructure);
                // structure.refStructures.push(CommonStructureConvertUtil.initFromObject(refTree));
            }));
        }
        //替换structuredTypes和refStructures中的propRefer
        ReferReplaceUtil.replacePropertyRefer(structure);
    };
    return CommonStructureConvertor;
}());
export { CommonStructureConvertor };
