// @farris/ui-footer/esm2015/lib/footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, HostBinding, Input } from '@angular/core';
import { FooterHeaderDirective } from './directive/footer-header.directive';
import { FooterHeaderToolbarExtendDirective } from './directive/footer-header-toolbar-extend.directive';
import { FooterHeaderContentDirective } from './directive/footer-header-content.directive';
import { FooterHeaderToolbarDirective } from './directive/footer-header-toolbar.directive';
import { LocaleService } from '@farris/ui-locale';
export class FooterComponent {
    /**
     * @param {?} localeService
     */
    constructor(localeService) {
        this.localeService = localeService;
        this.cls = true;
        this.textCollapse = this.localeService.getValue('footer.collapseText');
        this.textExpand = this.localeService.getValue('footer.expandText');
        // 显示头部区域
        this.showHeader = true;
        // 没有用模板单独定义class
        this.headerCls = "";
        // 单独定义主内容区的class
        this.contentCls = "";
        // 展开状态
        this.collapseStatus = true;
        // 是否启用展开收起
        this.enableAccordion = true;
    }
    // 展开
    /**
     * @return {?}
     */
    get getCollapseCls() {
        if (this.enableAccordion) {
            return this.collapseStatus;
        }
        else {
            return false;
        }
    }
    // 启用展开收起
    /**
     * @return {?}
     */
    get enableAccordionCls() {
        return this.enableAccordion;
    }
    // 收起文字
    /**
     * @param {?} value
     * @return {?}
     */
    set collapseText(value) {
        if (value) {
            this.textCollapse = value;
        }
    }
    /**
     * @return {?}
     */
    get collapseText() {
        return this.textCollapse;
    }
    // 展开文字
    /**
     * @param {?} value
     * @return {?}
     */
    set expandText(value) {
        if (value) {
            this.textExpand = value;
        }
    }
    /**
     * @return {?}
     */
    get expandText() {
        return this.textExpand;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    // 收起还是展开
    /**
     * @param {?} ev
     * @return {?}
     */
    expadnOrCollapse(ev) {
        ev && ev.stopImmediatePropagation();
        this.collapseStatus = !this.collapseStatus;
    }
}
FooterComponent.decorators = [
    { type: Component, args: [{
                selector: 'farris-footer',
                template: `
   <footer class="f-cmp-footer-container">
      <ng-container *ngIf="showHeader&&headerDirective">
          <header class="f-cmp-footer-header" [ngClass]="headerDirective.extendCls">
              <ng-template  [ngTemplateOutlet]="headerDirective.templateRef"></ng-template> 
          </header>
      </ng-container>
      <ng-container *ngIf="showHeader&&!headerDirective">
        <header class="f-cmp-footer-header" [ngClass]="headerCls">
          <div class="f-content" *ngIf="headerContentDirective" [ngClass]="headerContentDirective.extendCls">
            <ng-template [ngTemplateOutlet]="headerContentDirective.templateRef"></ng-template> 
          </div>
          <ng-container *ngIf="headerToolbarDirective">
              <div class="f-toolbar" [ngClass]="headerToolbarDirective.extendCls">
                  <ng-template  [ngTemplateOutlet]="headerToolbarDirective.templateRef"></ng-template> 
              </div>
          </ng-container>
          <div class="f-toolbar" *ngIf="!headerToolbarDirective">
            <ng-template *ngIf="headerToolbarExtendDirective" [ngTemplateOutlet]="headerToolbarExtendDirective.templateRef"></ng-template> 
            <div class="toolbar--collapse" *ngIf="enableAccordion" (click)="expadnOrCollapse($event)">
                <span class="toobar--collapse-text">
                   <ng-container  *ngIf="collapseStatus">{{textCollapse}}</ng-container>
                   <ng-container  *ngIf="!collapseStatus">{{textExpand}}</ng-container>
                </span>              
                <i class="f-icon toolbar--collapse-icon"></i>
            </div>
          </div>
        </header>
      </ng-container> 
      <div class="f-cmp-footer-content" [ngClass]="contentCls">
        <ng-content></ng-content>
      </div>
   </footer>
  `
            }] }
];
/** @nocollapse */
FooterComponent.ctorParameters = () => [
    { type: LocaleService }
];
FooterComponent.propDecorators = {
    cls: [{ type: HostBinding, args: ['class.f-cmp-footer',] }],
    getCollapseCls: [{ type: HostBinding, args: ['class.f-state-collapse',] }],
    enableAccordionCls: [{ type: HostBinding, args: ['class.f-cmp-footer-accordion',] }],
    collapseText: [{ type: Input }],
    expandText: [{ type: Input }],
    showHeader: [{ type: Input }],
    headerCls: [{ type: Input }],
    contentCls: [{ type: Input }],
    collapseStatus: [{ type: Input }],
    enableAccordion: [{ type: Input }],
    headerDirective: [{ type: ContentChild, args: [FooterHeaderDirective,] }],
    headerContentDirective: [{ type: ContentChild, args: [FooterHeaderContentDirective,] }],
    headerToolbarExtendDirective: [{ type: ContentChild, args: [FooterHeaderToolbarExtendDirective,] }],
    headerToolbarDirective: [{ type: ContentChild, args: [FooterHeaderToolbarDirective,] }]
};
if (false) {
    /** @type {?} */
    FooterComponent.prototype.cls;
    /** @type {?} */
    FooterComponent.prototype.textCollapse;
    /** @type {?} */
    FooterComponent.prototype.textExpand;
    /** @type {?} */
    FooterComponent.prototype.showHeader;
    /** @type {?} */
    FooterComponent.prototype.headerCls;
    /** @type {?} */
    FooterComponent.prototype.contentCls;
    /** @type {?} */
    FooterComponent.prototype.collapseStatus;
    /** @type {?} */
    FooterComponent.prototype.enableAccordion;
    /** @type {?} */
    FooterComponent.prototype.headerDirective;
    /** @type {?} */
    FooterComponent.prototype.headerContentDirective;
    /** @type {?} */
    FooterComponent.prototype.headerToolbarExtendDirective;
    /** @type {?} */
    FooterComponent.prototype.headerToolbarDirective;
    /**
     * @type {?}
     * @private
     */
    FooterComponent.prototype.localeService;
}
