// @farris/appointment-calendar/esm2015/lib/appointment-locales.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const APPOINTMENT_CALENDAR_LOCALES = {
    'en': {
        appointment: {
            today: 'Today',
            thisWeek: 'This week',
            day: 'Day',
            week: 'Week',
            prevDay: 'Prev day',
            prevWeek: 'Prev week',
            nextDay: 'Next day',
            nextWeek: 'Next week',
            dateFormat: 'yyyy-MM-dd',
            datePlace: 'Pleasy select a date.',
            weeks: { 一: 'Mon', 二: 'Tue', 三: 'Wed', 四: 'Thu', 五: 'Fri', 六: 'Sat', 日: 'Sun' },
            detail: {
                edit: 'Edit',
                del: 'Delete',
                close: 'Close',
                dateFormat: 'MM-dd'
            }
        }
    },
    'zh-CHS': {
        appointment: {
            today: '今天',
            thisWeek: '本周',
            day: '日',
            week: '周',
            prevDay: '前一天',
            prevWeek: '前一周',
            nextDay: '下一天',
            nextWeek: '下一周',
            dateFormat: 'yyyy年MM月dd日',
            datePlace: '请选择',
            weeks: { 一: '周一', 二: '周二', 三: '周三', 四: '周四', 五: '周五', 六: '周六', 日: '周日' },
            detail: {
                edit: '编辑',
                del: '删除',
                close: '关闭',
                dateFormat: 'MM月dd日'
            }
        }
    },
    'zh-CHT': {
        appointment: {
            today: '今天',
            thisWeek: '本周',
            day: '日',
            week: '周',
            prevDay: '前一天',
            prevWeek: '前一周',
            nextDay: '下一天',
            nextWeek: '下一周',
            dateFormat: 'yyyy年MM月dd日',
            datePlace: '請選擇',
            weeks: { 一: '周一', 二: '周二', 三: '周三', 四: '周四', 五: '周五', 六: '周六', 日: '周日' },
            detail: {
                edit: '編輯',
                del: '刪除',
                close: '關閉',
                dateFormat: 'MM月dd日'
            }
        }
    }
};
