// @gsp-sys/rtf-common/esm2015/lib/services/funcsservice/funcs.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { FrameworkVariableService } from '../frmVariableService/framework-variable.service';
export class FuncsService {
    /**
     * @param {?} frmVariableService
     */
    constructor(frmVariableService) {
        this.frmVariableService = frmVariableService;
        this.allgspfuncs = 'allgspfuncs';
        this.allfuncinvoks = 'allfuncinvoks';
        if (!this.frmVariableService) {
            this.frmVariableService = new FrameworkVariableService();
        }
    }
    /**
     * 缓存所有功能
     * @param {?} funcs
     * @return {?}
     */
    cacheAllFuncs(funcs) {
        this.frmVariableService.frameworkVariableReg(this.allgspfuncs, funcs);
    }
    /**
     * 获取所有缓存的功能
     * @return {?}
     */
    getAllCachedFuncs() {
        /** @type {?} */
        const funcsCache = this.frmVariableService.getFrmVariableByKey(this.allgspfuncs);
        return (/** @type {?} */ (funcsCache)) || [];
    }
    /**
     * 获取某功能
     * @param {?} funcId
     * @return {?}
     */
    getCachedFunc(funcId) {
        /** @type {?} */
        const allfuncs = this.getAllCachedFuncs();
        if (allfuncs) {
            return allfuncs.find((/**
             * @param {?} f
             * @return {?}
             */
            f => f.id === funcId));
        }
    }
    /**
     * @param {?} invoks
     * @return {?}
     */
    cacheAllFuncInvoks(invoks) {
        this.frmVariableService.frameworkVariableReg(this.allfuncinvoks, invoks);
    }
    /**
     * @return {?}
     */
    getAllCachedFuncInvoks() {
        /** @type {?} */
        const invokCache = this.frmVariableService.getFrmVariableByKey(this.allfuncinvoks);
        return (/** @type {?} */ (invokCache)) || [];
    }
}
FuncsService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FuncsService.ctorParameters = () => [
    { type: FrameworkVariableService, decorators: [{ type: Optional }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    FuncsService.prototype.allgspfuncs;
    /**
     * @type {?}
     * @private
     */
    FuncsService.prototype.allfuncinvoks;
    /**
     * @type {?}
     * @private
     */
    FuncsService.prototype.frmVariableService;
}
