// @farris/bef/esm2015/lib/utils/back_end_message.util.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/back_end_message.util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { AppContext, BackEndMessage, BACK_END_MESSAGE_HANDLER_TOKEN, NAMESPACE } from "@farris/devkit";
export class BackEndMessageUtil {
    /**
     * @param {?} messages
     * @param {?} injector
     * @param {?=} options
     * @return {?}
     */
    static handleMessage(messages, injector, options) {
        /** @type {?} */
        const namespace = injector.get(NAMESPACE, null);
        /** @type {?} */
        const appContext = injector.get(AppContext, null);
        /** @type {?} */
        const context = {
            ns: namespace,
            appContext
        };
        /** @type {?} */
        const bizMessages = this.buildBackEndMessages(messages);
        /** @type {?} */
        const backEndMessage = new BackEndMessage.Message(bizMessages, context);
        /** @type {?} */
        const backEndMessageHandler = injector.get(BACK_END_MESSAGE_HANDLER_TOKEN, null);
        if (backEndMessageHandler) {
            backEndMessageHandler.handle(backEndMessage, options);
        }
    }
    /**
     * @param {?} injector
     * @return {?}
     */
    static isBackEndMessageHandlerExist(injector) {
        /** @type {?} */
        const backEndMessageHandler = injector.get(BACK_END_MESSAGE_HANDLER_TOKEN, null);
        return backEndMessageHandler && true || false;
    }
    /**
     * @param {?} messages
     * @return {?}
     */
    static getFormlessMessages(messages) {
        /** @type {?} */
        const bizMessages = this.buildBackEndMessages(messages) || [];
        return bizMessages.filter((/**
         * @param {?} message
         * @return {?}
         */
        (message) => {
            return !message.location || !message.location.columns || message.location.columns.length < 1 || !message.location.nodeCode;
        }));
    }
    /**
     * @param {?} messages
     * @return {?}
     */
    static buildBackEndMessages(messages) {
        if (!messages || messages.length < 1 || !Array.isArray(messages)) {
            return null;
        }
        /** @type {?} */
        const bizMessages = [];
        messages.forEach((/**
         * @param {?} message
         * @return {?}
         */
        (message) => {
            /** @type {?} */
            const bizMessage = {
                level: message.level,
                message: message.message
            };
            if (message.hasOwnProperty('location') && message.location) {
                bizMessage.location = {
                    nodeCode: message.location.nodeCode,
                    rows: message.location.dataIds,
                    columns: message.location.columnNames
                };
            }
            bizMessages.push(bizMessage);
        }));
        return bizMessages;
    }
}
