// @farris/bef/esm2015/lib/utils/http_header.util.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/http_header.util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpHeaders } from '@angular/common/http';
class HttpHeaderUtil {
    /**
     * PlainObject => HttpHeaders
     * @param {?} headers
     * @return {?}
     */
    static convertToHeaders(headers) {
        if (!headers) {
            headers = new HttpHeaders();
        }
        // 已经是HttpHeaders，直接返回
        /** @type {?} */
        const isHttpHeaders = headers instanceof HttpHeaders;
        if (isHttpHeaders === true) {
            return headers;
        }
        if (Object.keys(headers).length > 0) {
            /** @type {?} */
            let httpHeaders = new HttpHeaders();
            Object.keys(headers).forEach((/**
             * @param {?} prop
             * @return {?}
             */
            prop => {
                httpHeaders = httpHeaders.append(prop, headers[prop]);
            }));
            headers = httpHeaders;
        }
        else {
            headers = new HttpHeaders();
        }
        return headers;
    }
    // #region 添加特定含义的header
    /**
     * 追加SessionId头
     * @param {?} headers HttpHeaders对象
     * @param {?} sessionId
     * @return {?}
     */
    static appendSessionId(headers, sessionId) {
        return headers.append('SessionId', sessionId);
    }
    /**
     * 追加CommonVariable头
     * @param {?} headers HttpHeaders对象
     * @param {?} commonVariable
     * @return {?}
     */
    static appendCafRuntimeCommonVariable(headers, commonVariable) {
        return headers.append('X-CAF-Runtime-CommonVariable', commonVariable);
    }
    /**
     * 追加X-CAF-Runtime-Context头
     * @param {?} headers HttpHeaders对象
     * @param {?} context 等价于BeSessionId
     * @return {?}
     */
    static appendCafRuntimeContext(headers, context) {
        return headers.append('X-CAF-Runtime-Context', context);
    }
    /**
     * 追加Content-Type头
     * @param {?} headers HttpHeaders对象
     * @param {?=} contentType
     * @return {?}
     */
    static appendContextType(headers, contentType) {
        contentType = contentType ? contentType : 'application/json';
        return headers.append('Content-Type', contentType);
    }
    /**
     * @param {?} headers
     * @param {?=} requireMessage
     * @return {?}
     */
    static appendRequireMessage(headers, requireMessage = true) {
        return headers.append('Bff-Require-Message', requireMessage.toString());
    }
    /**
     * 组合表单复用，表单实例id
     * @param {?} headers headers
     * @param {?} funcInstId 表单实例id
     * @return {?}
     */
    static appendFuncInstId(headers, funcInstId) {
        return headers.append('Func-Inst-Id', funcInstId);
    }
    /**
     * http headers to json
     * @param {?} headers headers
     * @return {?}
     */
    static toJson(headers) {
        if (!(headers instanceof HttpHeaders)) {
            return null;
        }
        /** @type {?} */
        const result = {};
        headers.keys().forEach((/**
         * @param {?} header
         * @return {?}
         */
        header => {
            /** @type {?} */
            const values = headers.getAll(header);
            if (values.length > 1) {
                result[header] = values;
            }
            else {
                result[header] = values[0] || '';
            }
        }));
        return result;
    }
}
export { HttpHeaderUtil };
