// @farris/bef/esm2015/lib/utils/request_info.util.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/request_info.util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class RequestInfoUtil {
    /**
     * 检查是否有requestInfo
     * @param {?} options
     * @return {?}
     */
    static hasRequestInfo(options) {
        if (!options || !options.body) {
            return false;
        }
        /** @type {?} */
        const body = options.body;
        // 有多个参数的时候
        if (body.requestInfo && body.requestInfo.dataChange) {
            return true;
        }
        if (body.dataChange) {
            return true;
        }
        return false;
    }
    /**
     * 检查是否有RquestInfo
     * \@summary
     * 产品部依赖了请求时传递RequestInfo，返回时直接取到returnValue的特性。
     * @param {?} options
     * @return {?}
     */
    static hasRequestInfoWithR(options) {
        if (!options || !options.body) {
            return false;
        }
        /** @type {?} */
        const body = options.body;
        // 有多个参数的时候
        if (body.RequestInfo && body.RequestInfo.dataChange) {
            return true;
        }
        if (body.dataChange) {
            return true;
        }
        return false;
    }
    /**
     * 构造RequestInfo
     * @param {?} repository
     * @return {?}
     */
    static buildRequestInfo(repository) {
        /** @type {?} */
        const requestInfo = {
            dataChange: repository.entityManager.buildAllEntityChangeDetails(),
            variableChange: repository.variableManager.buildChangeDetail()
        };
        return requestInfo;
    }
    /**
     * 构造带requestInfo的body（仅适用于body中有多个参数的情况）
     * \@summary
     * 一个参数时body=requestInfo，不适用该方法。
     * @param {?} body
     * @param {?} repository
     * @return {?}
     */
    static buildBodyWithRequestInfo(body, repository) {
        if (body.RequestInfo || body.requestInfo) {
            return body;
        }
        /** @type {?} */
        const requestInfo = this.buildRequestInfo(repository);
        /** @type {?} */
        const bodyWithRequestInfo = Object.assign({}, body, { requestInfo: requestInfo });
        return bodyWithRequestInfo;
    }
    /**
     * @param {?} options
     * @return {?}
     */
    static getRequestInfo(options) {
        if (!options || !options.body) {
            return null;
        }
        /** @type {?} */
        const body = options.body;
        // 有多个参数的时候
        if (body.requestInfo) {
            return body.requestInfo;
        }
        if (body.dataChange || body.variableChange) {
            return body;
        }
        return null;
    }
}
export { RequestInfoUtil };
