// @farris/component-querycondition/esm2015/lib/entity/conditionvalue/checkbox-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/checkbox-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
// CheckBoxValue初始化，与'app-solution-checkbox'组件初始化是分离的
// CheckBoxValue 要求value是布尔类型，组件要求value是true时，值应为true字符串
export class CheckBoxValue {
    /**
     * @return {?}
     */
    clearValue() {
        this.value = [];
    }
    // 如果组件的返回值是字符串，转化为布尔值
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        if (value.length) {
            this.value = value.map((/**
             * @param {?} v
             * @return {?}
             */
            v => {
                if (typeof v == 'string') {
                    return true;
                }
                else {
                    return v;
                }
            }));
        }
        else {
            this.value = value;
        }
    }
    // 把CheckBoxValue的true布尔值，转化为组件可识别的字符串
    /**
     * @return {?}
     */
    getOriginalValue() {
        if (this.value.length) {
            return this.value.map((/**
             * @param {?} v
             * @return {?}
             */
            v => {
                if (v === true) {
                    return "true";
                }
                else {
                    return v;
                }
            }));
        }
        return this.value;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.BoolCheck;
    }
    // CheckBoxValue 初始化，传入的值未做任何修改
    /**
     * @param {?=} value
     */
    constructor(value = []) {
        this.value = value;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return this.value.length == 0;
    }
}
if (false) {
    /** @type {?} */
    CheckBoxValue.prototype.value;
}
