// @farris/component-querycondition/esm2015/lib/entity/conditionvalue/single-number-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/single-number-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
export class SingleNumberValue {
    /**
     * @return {?}
     */
    isEmpty() {
        return this.numValue == null;
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.numValue = undefined;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        throw new Error("Method not implemented.");
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.numValue;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.SingleNumber;
    }
    /**
     * @param {?=} data
     */
    constructor(data = null) {
        this.numValue = data == null ? null : parseFloat(data);
    }
}
if (false) {
    /** @type {?} */
    SingleNumberValue.prototype.numValue;
}
