// @farris/component-querycondition/esm2015/lib/entity/conditionvalue/single-date-time-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/single-date-time-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
export class SingleDateTimeValue {
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        this.datetimeValue = value.formatted;
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.datetimeValue;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.SingleDateTime;
    }
    /**
     * @param {?=} datetime
     */
    constructor(datetime = '') {
        this.datetimeValue = datetime;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return !this.datetimeValue;
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.datetimeValue = undefined;
    }
}
if (false) {
    /** @type {?} */
    SingleDateTimeValue.prototype.datetimeValue;
}
