// @farris/component-querycondition/esm2015/lib/entity/conditionvalue/single-month-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/single-month-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
export class SingleMonthValue {
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        this.monthValue = value.formatted;
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.monthValue;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.SingleMonth;
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.monthValue = undefined;
    }
    /**
     * @param {?=} month
     */
    constructor(month = '') {
        this.monthValue = month;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return !this.monthValue;
    }
}
if (false) {
    /** @type {?} */
    SingleMonthValue.prototype.monthValue;
}
