// @farris/component-querycondition/esm2015/lib/entity/conditionvalue/single-year-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/single-year-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
export class SingleYearValue {
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        this.yearValue = value.formatted;
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.yearValue;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.SingleYear;
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.yearValue = undefined;
    }
    /**
     * @param {?=} year
     */
    constructor(year = '') {
        this.yearValue = year;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return !this.yearValue;
    }
}
if (false) {
    /** @type {?} */
    SingleYearValue.prototype.yearValue;
}
