// @farris/component-querycondition/esm2015/lib/entity/conditionvalue/textvalue.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/textvalue.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
export class TextValue {
    /**
     * @return {?}
     */
    clearValue() {
        this.value = undefined;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        this.value = value;
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.value;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.Text;
    }
    /**
     * @param {?=} data
     */
    constructor(data = { value: '' }) {
        // this.value = data.value ? data.value.trim() : '';
        this.value = data.value ? data.value : '';
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return !this.value;
    }
}
if (false) {
    /** @type {?} */
    TextValue.prototype.value;
}
