// @farris/mobile-command-services/esm2015/lib/map-services/static-map.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/map-services/static-map.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class StaticMap {
    constructor() {
        // 地图尺寸
        this.size = { width: 400, height: 400 };
        // 标记尺寸  三种
        this.markerSize = 'mid';
        // 标记数组
        this.markers = [];
    }
}
if (false) {
    /** @type {?} */
    StaticMap.prototype.key;
    /** @type {?} */
    StaticMap.prototype.location;
    /** @type {?} */
    StaticMap.prototype.zoom;
    /** @type {?} */
    StaticMap.prototype.size;
    /** @type {?} */
    StaticMap.prototype.scale;
    /** @type {?} */
    StaticMap.prototype.markerColor;
    /** @type {?} */
    StaticMap.prototype.markerSize;
    /** @type {?} */
    StaticMap.prototype.markers;
    /** @type {?} */
    StaticMap.prototype.labels;
    /** @type {?} */
    StaticMap.prototype.paths;
    /** @type {?} */
    StaticMap.prototype.traffic;
    /** @type {?} */
    StaticMap.prototype.sig;
}
class MapLocation {
}
if (false) {
    /** @type {?} */
    MapLocation.prototype.longtitude;
    /** @type {?} */
    MapLocation.prototype.latitude;
}
class MapSize {
}
if (false) {
    /** @type {?} */
    MapSize.prototype.width;
    /** @type {?} */
    MapSize.prototype.height;
}
class MapMarker {
}
if (false) {
    /** @type {?} */
    MapMarker.prototype.longtitude;
    /** @type {?} */
    MapMarker.prototype.latitude;
}
