// @gsp-wf/wf-process-editor/esm2015/lib/views/assign-scope/other-assign-activity-node-info.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Condition } from "./condition";
export class DisplayInfor {
    /**
     * @param {?} name
     * @param {?} id
     */
    constructor(name, id) {
        this.name = name;
        this.activityDefinitionId = id;
        this.required = false;
    }
}
if (false) {
    /** @type {?} */
    DisplayInfor.prototype.name;
    /** @type {?} */
    DisplayInfor.prototype.activityDefinitionId;
    /** @type {?} */
    DisplayInfor.prototype.condition;
    /** @type {?} */
    DisplayInfor.prototype.required;
}
export class OtherAssignActivityNodeInfo {
    /**
     * @param {?} id
     * @param {?} condition
     * @param {?} required
     */
    constructor(id, condition, required) {
        this.activityDefinitionId = id;
        this.condition = new Condition();
        this.condition.Value = condition;
        this.required = required;
    }
}
if (false) {
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.activityDefinitionId;
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.condition;
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.required;
}
