// @farris/component-querycondition/esm5/lib/entity/controltype.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/controltype.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var ControlType = {
    Text: 0,
    SingleDate: 1,
    SmartHelp: 2,
    DropDownList: 3,
    DateRange: 4,
    NumberRange: 5,
    SingleNumber: 6,
    SingleYear: 7,
    BoolCheck: 8,
    DateTimeRange: 9,
    SingleMonth: 10,
    MonthRange: 11,
    SingleDateTime: 12,
    ComboLookUp: 13,
    Radio: 14,
    InputGroup: 15,
};
export { ControlType };
ControlType[ControlType.Text] = 'Text';
ControlType[ControlType.SingleDate] = 'SingleDate';
ControlType[ControlType.SmartHelp] = 'SmartHelp';
ControlType[ControlType.DropDownList] = 'DropDownList';
ControlType[ControlType.DateRange] = 'DateRange';
ControlType[ControlType.NumberRange] = 'NumberRange';
ControlType[ControlType.SingleNumber] = 'SingleNumber';
ControlType[ControlType.SingleYear] = 'SingleYear';
ControlType[ControlType.BoolCheck] = 'BoolCheck';
ControlType[ControlType.DateTimeRange] = 'DateTimeRange';
ControlType[ControlType.SingleMonth] = 'SingleMonth';
ControlType[ControlType.MonthRange] = 'MonthRange';
ControlType[ControlType.SingleDateTime] = 'SingleDateTime';
ControlType[ControlType.ComboLookUp] = 'ComboLookUp';
ControlType[ControlType.Radio] = 'Radio';
ControlType[ControlType.InputGroup] = 'InputGroup';
