// @farris/devkit/esm5/lib/schema/form-metadata.js
var FormContent = /** @class */ (function () {
    function FormContent() {
    }
    return FormContent;
}());
export { FormContent };
var FormContentForDB = /** @class */ (function () {
    function FormContentForDB(content) {
        this.Id = content.Id;
        this.Code = content.Code;
        this.Name = content.Name;
        this.Contents = JSON.stringify(content.Contents);
    }
    return FormContentForDB;
}());
export { FormContentForDB };
var FormMetadaDataDom = /** @class */ (function () {
    function FormMetadaDataDom() {
    }
    return FormMetadaDataDom;
}());
export { FormMetadaDataDom };
var FormModule = /** @class */ (function () {
    function FormModule() {
    }
    return FormModule;
}());
export { FormModule };
// export class FormDeclaration {
//     events: Array<any>;
//     commands: Array<any>;
//     states: Array<any>;
// }
var FormOptions = /** @class */ (function () {
    function FormOptions() {
    }
    return FormOptions;
}());
export { FormOptions };
