// @qdp/ide-cmp/esm5/lib/component/charts-schema-editor/charts-schema-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Injector, HostBinding, ViewEncapsulation, Input, Output, EventEmitter, ViewChild } from '@angular/core';
import { ChartsTypeComponent } from './charts-type/charts-type.component';
import { ChartsDataComponent } from './charts-data/charts-data.component';
import { ChartsStyleComponent } from './charts-style/charts-style.component';
import { FarrisTabsComponent } from '@farris/ui-tabs';
import { LocalizeService } from '@qdp/localize';
var ChartsSchemaEditorComponent = /** @class */ (function () {
    function ChartsSchemaEditorComponent(injector) {
        this.injector = injector;
        this.cls = "f-utils-flex-column h-100";
        this.tabsSelectChange = new EventEmitter();
        this.localizeService = this.injector.get(LocalizeService);
    }
    /**
     * @return {?}
     */
    ChartsSchemaEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} data
     * @return {?}
     */
    ChartsSchemaEditorComponent.prototype.handleTabSelect = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        /** @type {?} */
        var option = this.getOption();
        this.tabsSelectChange.emit(option);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ChartsSchemaEditorComponent.prototype.chartTypeChange = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.chartsType = value;
    };
    /**
     * @return {?}
     */
    ChartsSchemaEditorComponent.prototype.getOption = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var option = {};
        if (this.chartsTypeCmp) {
            option['chartsType'] = this.chartsTypeCmp.getChartsTypeOption();
            // console.log("chartsType:");
            // console.log(option['chartsType']);
        }
        if (this.chartsDataCmp) {
            option['chartsData'] = this.chartsDataCmp.getChartsDataOption();
            option['groupSummarySet'] = this.chartsDataCmp.getGroupSummarySetOption();
            // console.log("chartsData:");
            // console.log(option['chartsData']);
        }
        if (this.chartsStyleCmp) {
            option['chartsStyle'] = this.chartsStyleCmp.getChartsStyleOption();
            option['customOption'] = this.chartsStyleCmp.getCustomOption();
            // console.log("chartsStyle:");
            // console.log(option['chartsStyle']);
        }
        return option;
    };
    ChartsSchemaEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'charts-schema-editor',
                    template: "<!--<kendo-tabstrip #tabStrip id=\"chartsSchemaEditor\" class=\"schema-editor-tab h-100 pt-1\" (tabSelect)=\"handleTabSelect($event)\">-->\r\n<farris-tabs  [value]=\"'type'\" [contentFill]=\"true\" [autoTitleWidth]=\"true\"  (tabChange)=\"handleTabSelect($event)\">\r\n  <farris-tab [title]=\"localizeService.getValue('ideCmp.chartsSchemaManager.chartsType.tab')\" [id] = \"'type'\">\r\n    <ng-template #content>\r\n      <charts-type [currSchema]=\"currSchema\" class=\"schema-editor-tab\" (chartTypeChange)=\"chartTypeChange($event)\" ></charts-type>\r\n    </ng-template>\r\n  </farris-tab>\r\n  <farris-tab [title]=\"localizeService.getValue('ideCmp.chartsSchemaManager.chartsData.tab')\" [id] = \"'data'\">\r\n    <ng-template #content>\r\n      <charts-data [currSchema]=\"currSchema\" class=\"schema-editor-tab\" [colList]=\"colList\" [chartsType]=\"chartsType\" ></charts-data>\r\n    </ng-template>\r\n  </farris-tab>\r\n  <farris-tab [title]=\"localizeService.getValue('ideCmp.chartsSchemaManager.chartsStyle.tab')\" [id] = \"'style'\">\r\n    <ng-template #content>\r\n      <charts-style [currSchema]=\"currSchema\" class=\"schema-editor-tab\" [chartsType]=\"chartsType\" ></charts-style>\r\n    </ng-template>\r\n  </farris-tab>\r\n</farris-tabs>\r\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".schema-editor-tab{position:absolute;top:0;bottom:0;left:0;right:0;overflow:hidden}"]
                }] }
    ];
    /** @nocollapse */
    ChartsSchemaEditorComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    ChartsSchemaEditorComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        currSchema: [{ type: Input }],
        colList: [{ type: Input }],
        controlType: [{ type: Input }],
        tabsSelectChange: [{ type: Output }],
        chartsTypeCmp: [{ type: ViewChild, args: [ChartsTypeComponent,] }],
        chartsDataCmp: [{ type: ViewChild, args: [ChartsDataComponent,] }],
        chartsStyleCmp: [{ type: ViewChild, args: [ChartsStyleComponent,] }],
        tabs: [{ type: ViewChild, args: [FarrisTabsComponent,] }]
    };
    return ChartsSchemaEditorComponent;
}());
export { ChartsSchemaEditorComponent };
if (false) {
    /** @type {?} */
    ChartsSchemaEditorComponent.prototype.cls;
    /** @type {?} */
    ChartsSchemaEditorComponent.prototype.currSchema;
    /** @type {?} */
    ChartsSchemaEditorComponent.prototype.colList;
    /** @type {?} */
    ChartsSchemaEditorComponent.prototype.controlType;
    /** @type {?} */
    ChartsSchemaEditorComponent.prototype.tabsSelectChange;
    /** @type {?} */
    ChartsSchemaEditorComponent.prototype.chartsTypeCmp;
    /** @type {?} */
    ChartsSchemaEditorComponent.prototype.chartsDataCmp;
    /** @type {?} */
    ChartsSchemaEditorComponent.prototype.chartsStyleCmp;
    /** @type {?} */
    ChartsSchemaEditorComponent.prototype.tabs;
    /** @type {?} */
    ChartsSchemaEditorComponent.prototype.chartsType;
    /** @type {?} */
    ChartsSchemaEditorComponent.prototype.localizeService;
    /**
     * @type {?}
     * @private
     */
    ChartsSchemaEditorComponent.prototype.injector;
}
