// @gspwidget/util/esm5/lib/repositories/datasrc.repo.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { addBaseUrl, DATA_PREFIX } from '../const';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
/** @type {?} */
export var datasrcUrls = {
    GET_BY_ID: DATA_PREFIX + "/DatasourceService/find/",
    // ?
    CREATE: DATA_PREFIX + "/DatasourceService/create",
    // 500
    DELETE: DATA_PREFIX + "/DatasourceService/delete",
    // ?
    GET_ALL: DATA_PREFIX + "/DatasourceService/findAll",
    // 500
    UPDATE: DATA_PREFIX + "/DatasourceService/update",
    // ?
    TRY: DATA_PREFIX + "/DatasourceService/try",
    GET_PARAMS_BY_PROVIDER_NAME: DATA_PREFIX + "/DataService/getDatasourceParameters/",
    GET_ALL_PROVIDERS: DATA_PREFIX + "/DataService/getProvicerNames",
};
addBaseUrl(datasrcUrls);
var DatasrcRepo = /** @class */ (function () {
    function DatasrcRepo(http) {
        this.http = http;
    }
    /**
     * @return {?}
     */
    DatasrcRepo.prototype.getDatasrcList = /**
     * @return {?}
     */
    function () {
        return this.http.post(datasrcUrls.GET_ALL, null);
    };
    /**
     * @param {?} id
     * @return {?}
     */
    DatasrcRepo.prototype.getDatasrcById = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http
            .get(datasrcUrls.GET_BY_ID + id);
    };
    /**
     * @param {?} datasrc
     * @return {?}
     */
    DatasrcRepo.prototype.saveDatasrc = /**
     * @param {?} datasrc
     * @return {?}
     */
    function (datasrc) {
        if (datasrc.id) {
            return this.http.post(datasrcUrls.UPDATE, datasrc);
            // TODO: 移走
            // .pipe(this.success())
        }
        else {
            return this.http.post(datasrcUrls.CREATE, datasrc);
            // TODO: 移走
            // .pipe(this.success())
        }
    };
    /**
     * @param {?} id
     * @return {?}
     */
    DatasrcRepo.prototype.deleteDatasrc = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http.post(datasrcUrls.DELETE, null, {
            params: { id: id }
        });
        // TODO: 移走
        // .pipe(this.success())
    };
    /**
     * @param {?} datasrc
     * @return {?}
     */
    DatasrcRepo.prototype.tryDatasrc = /**
     * @param {?} datasrc
     * @return {?}
     */
    function (datasrc) {
        return this.http.post(datasrcUrls.TRY, datasrc, { responseType: 'text' });
        // TODO: 移走
        // .pipe(tap(
        //   () => this.nzMsg.success('测试成功！'),
        //   () => this.nzMsg.error('测试失败！')
        // ))
    };
    /**
     * @return {?}
     */
    DatasrcRepo.prototype.getProviders = /**
     * @return {?}
     */
    function () {
        return this.http.get(datasrcUrls.GET_ALL_PROVIDERS);
    };
    /**
     * @param {?} providerName
     * @return {?}
     */
    DatasrcRepo.prototype.getParamsByProviderName = /**
     * @param {?} providerName
     * @return {?}
     */
    function (providerName) {
        return this.http.get(datasrcUrls.GET_PARAMS_BY_PROVIDER_NAME + providerName);
    };
    DatasrcRepo.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    DatasrcRepo.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    /** @nocollapse */ DatasrcRepo.ngInjectableDef = i0.defineInjectable({ factory: function DatasrcRepo_Factory() { return new DatasrcRepo(i0.inject(i1.HttpClient)); }, token: DatasrcRepo, providedIn: "root" });
    return DatasrcRepo;
}());
export { DatasrcRepo };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DatasrcRepo.prototype.http;
}
