// @gspwidget/util/esm5/lib/repositories/portlet.repo.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { addBaseUrl } from '../const';
import { EppPortlet } from '../types/portlet.type';
import { convertWidget } from './widget.repo';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
/** @type {?} */
export var portletUrls = {
    GET_BY_ID: "portlet/",
    GET_USER_ALL: "portlet/all",
    SAVE: "portlet/",
    GET_BY_MODULE_ID: "portlet/findByModuleId",
    DELETE_BY_ID: "portlet/",
    GET_PAGE: "page/findByPortletId"
};
/** @type {?} */
export var portletOps = {
    SAVE: "portletop/save?portletId=",
    GET_BY_PORTLET_ID: "portletop/getallpagerole?portletId="
};
addBaseUrl(portletUrls);
addBaseUrl(portletOps);
var PortletRepo = /** @class */ (function () {
    function PortletRepo(http) {
        this.http = http;
    }
    /**
     * @param {?} portletId
     * @return {?}
     */
    PortletRepo.prototype.getPageByPortletId = /**
     * @param {?} portletId
     * @return {?}
     */
    function (portletId) {
        return this.http.get(portletUrls.GET_PAGE, {
            params: { portletId: portletId }
        });
    };
    /**
     * @return {?}
     */
    PortletRepo.prototype.getPortlets = /**
     * @return {?}
     */
    function () {
        return this.http.get(portletUrls.GET_USER_ALL).pipe(map((/**
         * @param {?} portlets
         * @return {?}
         */
        function (portlets) { return portlets.map(convertPortlet); })));
    };
    /**
     * @param {?=} moduleId
     * @return {?}
     */
    PortletRepo.prototype.getPortletsByModuleId = /**
     * @param {?=} moduleId
     * @return {?}
     */
    function (moduleId) {
        return this.http.get(portletUrls.GET_BY_MODULE_ID, { params: moduleId ? { moduleId: moduleId } : null }).pipe(map((/**
         * @param {?} portlets
         * @return {?}
         */
        function (portlets) { return portlets.map(convertPortlet); })));
    };
    /**
     * @param {?} id
     * @return {?}
     */
    PortletRepo.prototype.getPortletById = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        // return this.storage.getPortletById(id)
        return this.http.get(portletUrls.GET_BY_ID + id).pipe(map(convertPortlet));
    };
    /**
     * @param {?} portlet
     * @return {?}
     */
    PortletRepo.prototype.savePortlet = /**
     * @param {?} portlet
     * @return {?}
     */
    function (portlet) {
        /** @type {?} */
        var serverPortlet = convertPortletInverse(portlet);
        return portlet.id
            ? this.http.post(portletUrls.SAVE, serverPortlet)
            : this.http.put(portletUrls.SAVE, serverPortlet);
    };
    /** portlet op */
    /**
     * portlet op
     * @param {?} portletId
     * @return {?}
     */
    PortletRepo.prototype.getPortletOps = /**
     * portlet op
     * @param {?} portletId
     * @return {?}
     */
    function (portletId) {
        return this.http.get(portletOps.GET_BY_PORTLET_ID + portletId);
    };
    /**
     * @param {?} portletId
     * @param {?} opIds
     * @return {?}
     */
    PortletRepo.prototype.savePortletOps = /**
     * @param {?} portletId
     * @param {?} opIds
     * @return {?}
     */
    function (portletId, opIds) {
        return this.http.post(portletOps.SAVE + portletId, opIds);
    };
    /**
     * @param {?} portletId
     * @return {?}
     */
    PortletRepo.prototype.deletePortletById = /**
     * @param {?} portletId
     * @return {?}
     */
    function (portletId) {
        return this.http.delete(portletUrls.DELETE_BY_ID + portletId);
    };
    PortletRepo.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    PortletRepo.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    /** @nocollapse */ PortletRepo.ngInjectableDef = i0.defineInjectable({ factory: function PortletRepo_Factory() { return new PortletRepo(i0.inject(i1.HttpClient)); }, token: PortletRepo, providedIn: "root" });
    return PortletRepo;
}());
export { PortletRepo };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PortletRepo.prototype.http;
}
/**
 * portlet 后端 -> 前端, 前端 -> 后端 转换
 * @param {?} serverPortlet
 * @return {?}
 */
function convertPortlet(serverPortlet) {
    serverPortlet.widget = convertWidget(serverPortlet.widget);
    if (serverPortlet.propValues)
        serverPortlet.propValues = JSON.parse(serverPortlet.propValues);
    return (/** @type {?} */ (Object.assign(new EppPortlet(null), serverPortlet)));
}
/**
 * @param {?} portlet
 * @return {?}
 */
function convertPortletInverse(portlet) {
    /** @type {?} */
    var serverPortlet = tslib_1.__assign({}, portlet);
    delete serverPortlet.widget;
    serverPortlet.propValues = (/** @type {?} */ (JSON.stringify(portlet.propValues)));
    return serverPortlet;
}
