// @gspwidget/util/esm5/lib/repositories/widget.repo.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { addBaseUrl } from '../const';
import { backendString2Json } from '../misc';
import { UtilService } from '../util.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../util.service";
/** @type {?} */
export var widgetUrls = {
    GET_BY_ID: "metaportlet/",
    GET_USER_ALL: "metaportlet/all",
    CREATE: "metaportlet/upload",
    EDIT: "metaportlet/editor",
    DELETE: "metaportlet/",
    GET_PORTLET: "portlet/findByMetaId",
};
addBaseUrl(widgetUrls);
/**
 * url 转换:
 * 服务端 url: /su/元部件名/文件路径
 * full url: host/runtime/web/metaportlet/su/元部件名/文件路径
 * @type {?}
 */
var WIDGET_PREFIX = location.origin + "/runtime/web/metaportlet"
/** widget 后端 -> 前端 转换 */
;
/**
 * widget 后端 -> 前端 转换
 * @param {?} widget
 * @return {?}
 */
export function convertWidget(widget) {
    widget.href = WIDGET_PREFIX + widget.href;
    widget.logoImgUrl = WIDGET_PREFIX + widget.logoImgUrl;
    widget.previewImgUrl = WIDGET_PREFIX + widget.previewImgUrl;
    widget.baseUrl = WIDGET_PREFIX + widget.baseUrl;
    widget.layout = widget.layout && JSON.parse(backendString2Json(widget.layout));
    return (/** @type {?} */ (widget));
}
var WidgetRepo = /** @class */ (function () {
    function WidgetRepo(http, util) {
        this.http = http;
        this.util = util;
    }
    /**
     * @return {?}
     */
    WidgetRepo.prototype.getWidgets = /**
     * @return {?}
     */
    function () {
        return this.http
            .get(widgetUrls.GET_USER_ALL)
            .pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return data.map(convertWidget); })));
    };
    /**
     * @param {?} id
     * @return {?}
     */
    WidgetRepo.prototype.getWidgetById = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http
            .get(widgetUrls.GET_BY_ID + id)
            .pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return convertWidget(data); })));
    };
    /**
     * @param {?} content
     * @return {?}
     */
    WidgetRepo.prototype.uploadWidget = /**
     * @param {?} content
     * @return {?}
     */
    function (content) {
        return this.util.uploadFile(widgetUrls.CREATE, content);
    };
    /**
     * @param {?} content
     * @return {?}
     */
    WidgetRepo.prototype.editorWidget = /**
     * @param {?} content
     * @return {?}
     */
    function (content) {
        return this.util.uploadFile(widgetUrls.EDIT, content);
    };
    /**
     * @param {?} id
     * @return {?}
     */
    WidgetRepo.prototype.deleteWidget = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http.delete(widgetUrls.DELETE + id, {});
    };
    /**
     * @param {?} id
     * @return {?}
     */
    WidgetRepo.prototype.findPortletByMeta = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http.get(widgetUrls.GET_PORTLET, {
            params: { id: id }
        });
    };
    WidgetRepo.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    WidgetRepo.ctorParameters = function () { return [
        { type: HttpClient },
        { type: UtilService }
    ]; };
    /** @nocollapse */ WidgetRepo.ngInjectableDef = i0.defineInjectable({ factory: function WidgetRepo_Factory() { return new WidgetRepo(i0.inject(i1.HttpClient), i0.inject(i2.UtilService)); }, token: WidgetRepo, providedIn: "root" });
    return WidgetRepo;
}());
export { WidgetRepo };
if (false) {
    /**
     * @type {?}
     * @private
     */
    WidgetRepo.prototype.http;
    /**
     * @type {?}
     * @private
     */
    WidgetRepo.prototype.util;
}
