// @gsp-dip/data-imp-exp/esm2015/lib/data-imp-exp.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { DropDownsModule } from '@progress/kendo-angular-dropdowns';
import { GridModule, PDFModule, ExcelModule } from '@progress/kendo-angular-grid';
import { ModalModule } from '@farris/ui-modal';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { LayoutModule as KendoLayoutModule } from '@progress/kendo-angular-layout';
import { TreeViewModule } from '@progress/kendo-angular-treeview';
import { FilterModule } from '@farris/ui-filter-editor';
import { FarrisTabsModule } from '@farris/ui-tabs';
import { PopoverModule } from '@farris/ui-popover';
import { ElementSelectComponent } from './element-select/element-select.component';
import { ImportDialogComponent } from './import-dialog/import-dialog.component';
import { ExportDialogComponent } from './custom-export-dialog/export-dialog.component';
import { CommonModule } from '@angular/common';
import { ValidateFormComponent } from './validate-form/validate-form.component';
import { FileDropDirective } from './service/UploaderService/FileUpload/file-drop.directive';
import { FileSelectDirective } from './service/UploaderService/FileUpload/file-selecct.directive';
import { DataExportService } from './service/DataIEService/data-export.service';
import { FileOperateService } from './service/DataIEService/file-operate.service';
import { CustomImportDialogComponent } from './custom-import-dialog/custom-import-dialog.component';
import { DataImportService } from './service/DataIEService/data-import.service';
import { CustomDataIEService } from './service/DataIEService/custom-data-ie.service';
import { ExportCustomSelectComponent } from './export-custom-select/export-custom-select.component';
import { DataieHandcraftService } from './service/DataIEService/dataie-handcraft.service';
import { RuleSelectComponent } from './rule-select/rule-select.component';
import { DatagridModule } from '@farris/ui-datagrid';
import { InputGroupModule } from '@farris/ui-input-group';
import { MultiSelectModule } from '@farris/ui-multi-select';
import { ImportDialogHandcraftComponent } from './import-dialog-handcraft/import-dialog-handcraft.component';
import { ComboListModule } from '@farris/ui-combo-list';
import { FarrisButtonModule } from '@farris/ui-button';
import { FarrisTooltipModule } from '@farris/ui-tooltip';
import { ProgressModule } from '@farris/ui-progress';
import { FarrisFormsModule } from '@farris/ui-forms';
import { ExtendComponent9Component } from './bakcomponents/extend-component9/extend-component9.component';
import { ExtendComponent8Component } from './bakcomponents/extend-component8/extend-component8.component';
import { ExtendComponent7Component } from './bakcomponents/extend-component7/extend-component7.component';
import { ExtendComponent6Component } from './bakcomponents/extend-component6/extend-component6.component';
import { ExtendComponent5Component } from './bakcomponents/extend-component5/extend-component5.component';
import { ExtendComponent4Component } from './bakcomponents/extend-component4/extend-component4.component';
import { ExtendComponent3Component } from './bakcomponents/extend-component3/extend-component3.component';
import { ExtendComponent2Component } from './bakcomponents/extend-component2/extend-component2.component';
import { ExtendComponent1Component } from './bakcomponents/extend-component1/extend-component1.component';
import { RuleCustomSettingComponent } from './bakcomponents/rule-custom-setting/rule-custom-setting.component';
import { DataieProgressComponent } from './bakcomponents/dataie-progress/dataie-progress.component';
import { RuleIdHelpComponent } from './design-time/rule-id-help/rule-id-help.component';
import { DataieRuleHelpComponent } from './design-time/dataie-rule-help/dataie-rule-help.component';
import { ExportCustomSelectHandcraftComponent } from './export-custom-select-handcraft/export-custom-select-handcraft.component';
import { LayoutModule } from '@farris/ui-layout';
import { DatagridFilterRowModule } from '@farris/ui-datagrid-filter';
import { FarrisSectionModule } from '@farris/ui-section';
export class DataImpExpModule {
}
DataImpExpModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    ElementSelectComponent,
                    ImportDialogComponent,
                    ExportDialogComponent,
                    ValidateFormComponent,
                    FileDropDirective,
                    FileSelectDirective,
                    CustomImportDialogComponent,
                    ExportCustomSelectComponent,
                    RuleSelectComponent,
                    ImportDialogHandcraftComponent,
                    ExportCustomSelectHandcraftComponent,
                    //extendcomponent
                    DataieProgressComponent,
                    RuleCustomSettingComponent,
                    ExtendComponent1Component,
                    ExtendComponent2Component,
                    ExtendComponent3Component,
                    ExtendComponent4Component,
                    ExtendComponent5Component,
                    ExtendComponent6Component,
                    ExtendComponent7Component,
                    ExtendComponent8Component,
                    ExtendComponent9Component,
                    RuleIdHelpComponent,
                    DataieRuleHelpComponent,
                ],
                exports: [
                    ElementSelectComponent,
                    ImportDialogComponent,
                    ExportDialogComponent,
                    ValidateFormComponent,
                    CustomImportDialogComponent,
                    ExportCustomSelectComponent,
                    ImportDialogHandcraftComponent,
                    FileDropDirective,
                    FileSelectDirective,
                    RuleIdHelpComponent,
                    DataieRuleHelpComponent
                ],
                entryComponents: [ImportDialogComponent, ImportDialogHandcraftComponent,
                    ExportDialogComponent, ValidateFormComponent, RuleSelectComponent,
                    CustomImportDialogComponent, ExportCustomSelectComponent,
                    RuleIdHelpComponent, DataieRuleHelpComponent, ExportCustomSelectHandcraftComponent,
                    //extendcomponent
                    DataieProgressComponent,
                    RuleCustomSettingComponent,
                    ExtendComponent1Component,
                    ExtendComponent2Component,
                    ExtendComponent3Component,
                    ExtendComponent4Component,
                    ExtendComponent5Component,
                    ExtendComponent6Component,
                    ExtendComponent7Component,
                    ExtendComponent8Component,
                    ExtendComponent9Component
                ],
                imports: [
                    CommonModule,
                    DropDownsModule,
                    GridModule,
                    ModalModule.forRoot(),
                    FormsModule,
                    FarrisFormsModule,
                    FarrisDialogModule,
                    KendoLayoutModule,
                    ExcelModule,
                    TreeViewModule,
                    FilterModule,
                    FarrisTabsModule,
                    PDFModule,
                    DatagridModule,
                    DatagridFilterRowModule.forRoot(),
                    InputGroupModule,
                    MultiSelectModule,
                    ComboListModule,
                    FarrisButtonModule,
                    PopoverModule,
                    FarrisTooltipModule,
                    ProgressModule,
                    LayoutModule,
                    FarrisSectionModule
                ],
                providers: [DataExportService, FileOperateService, DataImportService, CustomDataIEService, DataieHandcraftService
                ],
                bootstrap: []
            },] }
];
