// @gsp-lcm/bo-creator/fesm5/gsp-lcm-bo-creator.js
import { CommonModule } from '@angular/common';
import { Injectable, Component, Output, EventEmitter, Input, ViewChild, NgModule } from '@angular/core';
import { InputGroupModule } from '@farris/ui-input-group';
import { FormsModule } from '@angular/forms';
import { InputsModule } from '@progress/kendo-angular-inputs';
import { NotifyService, NotifyModule } from '@farris/ui-notify';
import { ComboListModule } from '@farris/ui-combo-list';
import { BusinessObjectEntity, BusinessObjectDtService, BoDtServiceModule } from '@gsp-lcm/bo-dt-service';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/views/bo-creator/common/validation-result.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ValidatorResult = /** @class */ (function () {
    function ValidatorResult() {
        this.illegal = false;
    }
    return ValidatorResult;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/views/bo-creator/utils/bo-create-utils.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BoCreatorUtils = /** @class */ (function () {
    function BoCreatorUtils() {
    }
    /**
     * @return {?}
     */
    BoCreatorUtils.prototype.newGuid = /**
     * @return {?}
     */
    function () {
        return (this.s4() + this.s4() + "-" + this.s4() + "-" + this.s4() + "-" + this.s4() + "-" + this.s4() + this.s4() + this.s4());
    };
    /**
     * @return {?}
     */
    BoCreatorUtils.prototype.s4 = /**
     * @return {?}
     */
    function () {
        return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
    };
    /**
     * @param {?} entities
     * @param {?} parentID
     * @return {?}
     */
    BoCreatorUtils.prototype.getMaxOrder = /**
     * @param {?} entities
     * @param {?} parentID
     * @return {?}
     */
    function (entities, parentID) {
        /** @type {?} */
        var maxNum = 0;
        /** @type {?} */
        var children = entities.filter((/**
         * @param {?} entity
         * @return {?}
         */
        function (entity) { return entity.parentID === parentID; }));
        if (children === undefined || children === null || children.length <= 0) {
            return maxNum;
        }
        else {
            children.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                if (element.sortOrder > maxNum) {
                    maxNum = element.sortOrder;
                }
            }));
        }
        return maxNum;
    };
    BoCreatorUtils.decorators = [
        { type: Injectable }
    ];
    return BoCreatorUtils;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/views/bo-creator/modal/app-modal/app-creator-modal.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var AppCreatorModalComponent = /** @class */ (function () {
    function AppCreatorModalComponent(boCreatorUtils, boDtService, notifyService) {
        this.boCreatorUtils = boCreatorUtils;
        this.boDtService = boDtService;
        this.notifyService = notifyService;
        this.appConfirmEmitter = new EventEmitter();
        this.currentBo = new BusinessObjectEntity();
        this.apps = [];
    }
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initData();
    };
    /**
     * @private
     * @return {?}
     */
    AppCreatorModalComponent.prototype.initData = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.boDtService.getBusinessObjects().subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (!data) {
                return;
            }
            //二级节点是APP
            _this.apps = data.filter((/**
             * @param {?} x
             * @return {?}
             */
            function (x) { return x.layer === 2; }));
            //一级节点是产品，目前默认一级节点（根节点）只有一个
            _this.parent = data.find((/**
             * @param {?} x
             * @return {?}
             */
            function (x) { return x.layer === 1; }));
        }));
    };
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.create = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var bo = this.getCurrentBo();
        if (!bo) {
            return;
        }
        this.boDtService.save(bo).subscribe((/**
         * @return {?}
         */
        function () {
            _this.appConfirmEmitter.emit(bo);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.notifyService.error('创建失败');
            throw error;
        }));
    };
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.getCurrentBo = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = this.appInfoValidator();
        if (result.illegal) {
            this.notifyService.warning(result.message);
            return null;
        }
        this.currentBo.id = this.boCreatorUtils.newGuid();
        this.currentBo.layer = 2;
        this.currentBo.parentID = this.parent.id;
        this.currentBo.sysInit = '0';
        this.currentBo.isSysInit = false;
        this.currentBo.isDetail = '0';
        /** @type {?} */
        var max = this.boCreatorUtils.getMaxOrder(this.apps, this.currentBo.parentID);
        this.currentBo.sortOrder = max + 1;
        return Object.assign(new BusinessObjectEntity(), this.currentBo);
    };
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.appInfoValidator = /**
     * @return {?}
     */
    function () {
        //新增信息空校验
        /** @type {?} */
        var result = this.emptyValidator();
        if (result.illegal) {
            return result;
        }
        //编号，名称命名规范性校验
        result = this.namingConventionValidator();
        if (result.illegal) {
            return result;
        }
        //APP编号重复校验
        return this.repeatValidator();
    };
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.emptyValidator = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = new ValidatorResult();
        if (!this.currentBo.code) {
            result.illegal = true;
            result.message = "\u5173\u952E\u5E94\u7528\u7684\u7F16\u53F7\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        if (!this.currentBo.name) {
            result.illegal = true;
            result.message = "\u5173\u952E\u5E94\u7528\u7684\u540D\u79F0\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        return result;
    };
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.namingConventionValidator = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = new ValidatorResult();
        /** @type {?} */
        var codeRegx = new RegExp('^[a-zA-Z\x7f-\xff][a-zA-Z0-9\x7f-\xff]*$');
        if (!codeRegx.test(this.currentBo.code)) {
            result.illegal = true;
            result.message = '关键应用的编号只能包含字母,数字,并且不可以用数字开头!';
            return result;
        }
        /** @type {?} */
        var nameRegx = new RegExp('[`~!@#$^&*()=|{}\':;\',\\[\\].<>/?~！@#￥……&*（）——|{}【】‘；：”“\'。，、？]');
        if (nameRegx.test(this.currentBo.name)) {
            result.illegal = true;
            result.message = '关键应用的名称包含非法字符，请重新输入!';
            return result;
        }
        return result;
    };
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.repeatValidator = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var result = new ValidatorResult();
        /** @type {?} */
        var filterApp = this.apps.find((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.code.toLowerCase() === _this.currentBo.code.toLowerCase(); }));
        if (filterApp) {
            result.illegal = true;
            result.message = "\u5173\u952E\u5E94\u7528\u7684\u7F16\u53F7" + this.currentBo.code + "\u5DF2\u5B58\u5728";
            return result;
        }
        return result;
    };
    AppCreatorModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-app-creator-modal',
                    template: "\r\n<div class=\"cloumn\" style=\"margin-top: 28px;\">\r\n\r\n  <div class=\"app-Info row\">\r\n    <label class=\"col-form-label ps-col-form-label\">\r\n      <span class=\"farris-label-info text-danger\">*</span>\r\n      <span class=\"farris-label-text\">\u5E94\u7528\u7F16\u53F7</span>\r\n    </label>\r\n    <div class=\"col-9\">\r\n      <input-group [(ngModel)]=\"currentBo.code\"  autocomplete=\"off\" name=\"code\" ></input-group>\r\n    </div>\r\n  </div>\r\n\r\n  <div class=\"app-Info row\">\r\n    <label class=\"col-form-label ps-col-form-label\">\r\n      <span class=\"farris-label-info text-danger\">*</span>\r\n      <span class=\"farris-label-text\">\u5E94\u7528\u540D\u79F0</span>\r\n    </label>\r\n    <div class=\"col-9\">\r\n      <input-group [(ngModel)]=\"currentBo.name\"  autocomplete=\"off\" name=\"name\" ></input-group>\r\n    </div>\r\n  </div>\r\n\r\n  <div class=\"app-Info row\">\r\n    <label class=\"col-form-label ps-col-form-label\">\r\n      <span class=\"farris-label-text\">\u5E94\u7528\u63CF\u8FF0</span>\r\n    </label>\r\n    <div class=\"col-9\">\r\n      <input-group [(ngModel)]=\"currentBo.description\"  autocomplete=\"off\" name=\"description\"></input-group>\r\n    </div>\r\n  </div>\r\n\r\n</div>",
                    styles: [".app-Info{margin:0 0 16px 20px}div .ps-col-form-label{width:94px;text-align:right}"]
                }] }
    ];
    /** @nocollapse */
    AppCreatorModalComponent.ctorParameters = function () { return [
        { type: BoCreatorUtils },
        { type: BusinessObjectDtService },
        { type: NotifyService }
    ]; };
    AppCreatorModalComponent.propDecorators = {
        appConfirmEmitter: [{ type: Output }]
    };
    return AppCreatorModalComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/views/bo-creator/modal/su-modal/su-creator-modal.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SuCreatorModalComponent = /** @class */ (function () {
    function SuCreatorModalComponent(boCreatorUtils, boDtService, notifyService) {
        this.boCreatorUtils = boCreatorUtils;
        this.boDtService = boDtService;
        this.notifyService = notifyService;
        this.suConfirmEmitter = new EventEmitter();
        /**
         * 当前BO节点
         */
        this.currentBo = new BusinessObjectEntity();
        /**
         * 关键应用集合
         */
        this.items = [];
        /**
         * 选择应用节点下的所有服务单元集合
         */
        this.suList = [];
        /**
         * 所有业务对象节点集合
         */
        this.data = [];
        this.parent = new BusinessObjectEntity();
        this.labelLength = '2';
        this.inputLength = '9';
    }
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initData();
    };
    /**
     * @private
     * @return {?}
     */
    SuCreatorModalComponent.prototype.initData = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.boDtService.getBusinessObjects().subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.data = data;
            _this.items = data.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.layer === 2; }));
            _this.suList = data.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.layer === 3; }));
        }));
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SuCreatorModalComponent.prototype.appSelectChanged = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.parent = $event.data;
    };
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.create = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var bo = this.getCurrentBo();
        if (!bo) {
            return;
        }
        this.boDtService.save(bo).subscribe((/**
         * @return {?}
         */
        function () {
            _this.suConfirmEmitter.emit(bo);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.notifyService.error('创建失败');
            throw error;
        }));
    };
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.getCurrentBo = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = this.suInfoValidator();
        if (result.illegal) {
            this.notifyService.warning(result.message);
            return null;
        }
        this.currentBo.id = this.boCreatorUtils.newGuid();
        this.currentBo.layer = 3;
        this.currentBo.parentID = this.parentId;
        this.currentBo.sysInit = '0';
        this.currentBo.isSysInit = false;
        this.currentBo.isDetail = '0';
        /** @type {?} */
        var max = this.boCreatorUtils.getMaxOrder(this.suList, this.currentBo.parentID);
        this.currentBo.sortOrder = max + 1;
        return Object.assign(new BusinessObjectEntity(), this.currentBo);
    };
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.suInfoValidator = /**
     * @return {?}
     */
    function () {
        //新增信息空校验
        /** @type {?} */
        var result = this.emptyValidator();
        if (result.illegal) {
            return result;
        }
        //编号，名称命名规范性校验
        result = this.namingConventionValidator();
        if (result.illegal) {
            return result;
        }
        //APP编号重复校验
        return this.repeatValidator();
    };
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.emptyValidator = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = new ValidatorResult();
        if (!this.parentId) {
            result.illegal = true;
            result.message = "\u5173\u952E\u5E94\u7528\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        if (!this.currentBo.code) {
            result.illegal = true;
            result.message = "\u670D\u52A1\u5355\u5143\u7684\u7F16\u53F7\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        if (!this.currentBo.name) {
            result.illegal = true;
            result.message = "\u670D\u52A1\u5355\u5143\u7684\u540D\u79F0\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        return result;
    };
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.namingConventionValidator = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = new ValidatorResult();
        /** @type {?} */
        var codeRegx = new RegExp('^[a-zA-Z\x7f-\xff][a-zA-Z0-9\x7f-\xff]*$');
        if (!codeRegx.test(this.currentBo.code)) {
            result.illegal = true;
            result.message = '服务单元的编号只能包含字母,数字,并且不可以用数字开头!';
            return result;
        }
        /** @type {?} */
        var nameRegx = new RegExp('[`~!@#$^&*()=|{}\':;\',\\[\\].<>/?~！@#￥……&*（）——|{}【】‘；：”“\'。，、？]');
        if (nameRegx.test(this.currentBo.name)) {
            result.illegal = true;
            result.message = '服务单元的名称包含非法字符，请重新输入!';
            return result;
        }
        return result;
    };
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.repeatValidator = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var result = new ValidatorResult();
        if (!this.suList || this.suList.length <= 0) {
            return result;
        }
        /** @type {?} */
        var repeatCodeBO = this.suList.find((/**
         * @param {?} bo
         * @return {?}
         */
        function (bo) { return bo.code.toLowerCase() === _this.currentBo.code.toLowerCase() && bo.layer == 3; }));
        if (repeatCodeBO) {
            result.illegal = true;
            result.message = '微服务的编号不能与已有微服务编号重复，编号为：' + this.currentBo.code;
            return result;
        }
        return result;
    };
    SuCreatorModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-su-creator-modal',
                    template: "<div class=\"cloumn\" style=\"margin-top: 28px;\">\r\n  <div class=\"row suInfo\">\r\n    <label class=\"col-form-label ps-col-form-label\">\r\n      <span class=\"farris-label-info text-danger\">*</span>\r\n      <span class=\"farris-label-text\">\u5173\u952E\u5E94\u7528</span>\r\n    </label>\r\n    <div class=\"col-{{inputLength}}\">\r\n      <farris-combo-list [(ngModel)]=\"parentId\" [idField]=\"'id'\" [textField]=\"'name'\" [data]=\"items\"\r\n        (selectChange)=\"appSelectChanged($event)\" [panelHeight]=\"280\" [editable]=\"false\"></farris-combo-list>\r\n    </div>\r\n  </div>\r\n\r\n  <div class=\"row suInfo\">\r\n    <!-- <div class=\"col-{{labelLength}}\">\u7F16\u53F7</div> -->\r\n    <label class=\"col-form-label ps-col-form-label\">\r\n      <span class=\"farris-label-info text-danger\">*</span>\r\n      <span class=\"farris-label-text\">\u7F16\u53F7</span>\r\n    </label>\r\n    <div class=\"col-{{inputLength}}\">\r\n      <input-group [(ngModel)]=\"currentBo.code\" #boCode=\"ngModel\" required></input-group>\r\n    </div>\r\n    <!-- <div [hidden]=\"boCode.valid || boCode.pristine\" class=\"alert-danger\">\r\n      \u7F16\u53F7\u4E0D\u80FD\u4E3A\u7A7A\r\n    </div> -->\r\n  </div>\r\n  <div class=\"row suInfo\">\r\n    <label class=\"col-form-label ps-col-form-label\">\r\n      <span class=\"farris-label-info text-danger\">*</span>\r\n      <span class=\"farris-label-text\">\u540D\u79F0</span>\r\n    </label>\r\n    <!-- <div class=\"col-{{labelLength}}\">\u540D\u79F0</div> -->\r\n    <div class=\"col-{{inputLength}}\">\r\n      <input-group [(ngModel)]=\"currentBo.name\" #boName=\"ngModel\" required></input-group>\r\n    </div>\r\n    <!-- <div [hidden]=\"boName.valid || boName.pristine\" class=\"alert-danger\">\r\n      \u540D\u79F0\u4E0D\u80FD\u4E3A\u7A7A\r\n    </div> -->\r\n  </div>\r\n\r\n</div>",
                    styles: [".sucreator-wrapper{background:rgba(255,255,255,.8);width:95%;margin:40px auto 0;max-width:480px;padding:24px 18px;border-radius:8px;position:relative;z-index:100}.suInfo{margin:0 0 16px 20px}.sucreator-wrapper .col-form-label,.sucreator-wrapper .form-control{padding-top:3px;padding-bottom:3px}.sucreator-wrapper .col-form-label{width:70px;margin-right:12px;text-align:right;flex-shrink:0}.bocreator{background:#fff;box-shadow:0 2px 16px 0 rgba(0,0,0,.1);border-radius:4px}.boCreatorPanel{background-color:#fff}div .ps-col-form-label{width:94px;text-align:right}"]
                }] }
    ];
    /** @nocollapse */
    SuCreatorModalComponent.ctorParameters = function () { return [
        { type: BoCreatorUtils },
        { type: BusinessObjectDtService },
        { type: NotifyService }
    ]; };
    SuCreatorModalComponent.propDecorators = {
        parentId: [{ type: Input }],
        suConfirmEmitter: [{ type: Output }],
        appSelector: [{ type: ViewChild, args: ["appSelector",] }]
    };
    return SuCreatorModalComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/views/bo-creator/modal/bo-modal/bo-creator-modal.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BoCreatorModalComponent = /** @class */ (function () {
    function BoCreatorModalComponent(boCreatorUtils, boDtService, notifyService) {
        this.boCreatorUtils = boCreatorUtils;
        this.boDtService = boDtService;
        this.notifyService = notifyService;
        this.boConfirmEmitter = new EventEmitter();
        /**
         * 当前BO节点
         */
        this.currentBo = new BusinessObjectEntity();
        /**
         * 关键应用集合
         */
        this.items = [];
        /**
         * 选择应用节点下的所有服务单元集合
         */
        this.suList = [];
        /**
         * 所有业务对象节点集合
         */
        this.data = [];
        this.parentBo = new BusinessObjectEntity();
        this.labelLength = '2';
        this.inputLength = '9';
    }
    Object.defineProperty(BoCreatorModalComponent.prototype, "_parentId", {
        set: /**
         * @param {?} parentId
         * @return {?}
         */
        function (parentId) {
            this.parentId = parentId;
            this.getParentBo();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initData();
    };
    /**
     * @private
     * @return {?}
     */
    BoCreatorModalComponent.prototype.initData = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.boDtService.getBusinessObjects().subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.data = data;
            _this.items = data.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.parentID === _this.parentId && item.layer === 4 && item.isDetail === '0'; }));
            /** @type {?} */
            var children = data.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.parentID === _this.parentId; }));
            /** @type {?} */
            var boInSU = [];
            children.forEach((/**
             * @param {?} s
             * @return {?}
             */
            function (s) {
                if (s.isDetail == '0') {
                    boInSU.push(s);
                    /** @type {?} */
                    var bos = data.filter((/**
                     * @param {?} t
                     * @return {?}
                     */
                    function (t) { return t.parentID === s.id; }));
                    if (bos) {
                        bos.forEach((/**
                         * @param {?} bo
                         * @return {?}
                         */
                        function (bo) {
                            boInSU.push(bo);
                        }));
                    }
                }
                else {
                    boInSU.push(s);
                }
            }));
            _this.bosInSU = boInSU;
        }));
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.create = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var bo = this.getCurrentBo();
        if (!bo) {
            return;
        }
        this.boDtService.save(bo).subscribe((/**
         * @return {?}
         */
        function () {
            _this.boConfirmEmitter.emit(bo);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.notifyService.error('创建失败');
            throw error;
        }));
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.getCurrentBo = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = this.boInfoValidator();
        if (result.illegal) {
            this.notifyService.warning(result.message);
            return null;
        }
        this.currentBo.id = this.boCreatorUtils.newGuid();
        this.currentBo.layer = this.parentBo ? this.parentBo.layer + 1 : 4;
        this.currentBo.parentID = this.parentBo && this.parentBo.id ? this.parentBo.id : this.parentId;
        this.currentBo.sysInit = '0';
        this.currentBo.isSysInit = false;
        this.currentBo.isDetail = '1';
        /** @type {?} */
        var max = this.boCreatorUtils.getMaxOrder(this.bosInSU, this.currentBo.parentID);
        this.currentBo.sortOrder = max + 1;
        return Object.assign(new BusinessObjectEntity(), this.currentBo);
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.boInfoValidator = /**
     * @return {?}
     */
    function () {
        //新增信息空校验
        /** @type {?} */
        var result = this.emptyValidator();
        if (result.illegal) {
            return result;
        }
        //编号，名称命名规范性校验
        result = this.namingConventionValidator();
        if (result.illegal) {
            return result;
        }
        //APP编号重复校验
        return this.repeatValidator();
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.emptyValidator = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = new ValidatorResult();
        if (!this.suCode) {
            result.illegal = true;
            result.message = "\u6240\u5C5E\u5FAE\u670D\u52A1\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        if (!this.currentBo.code) {
            result.illegal = true;
            result.message = "\u4E1A\u52A1\u5BF9\u8C61\u7684\u7F16\u53F7\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        if (!this.currentBo.name) {
            result.illegal = true;
            result.message = "\u4E1A\u52A1\u5BF9\u8C61\u7684\u540D\u79F0\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        return result;
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.namingConventionValidator = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = new ValidatorResult();
        /** @type {?} */
        var codeRegx = new RegExp('^[a-zA-Z\x7f-\xff][a-zA-Z0-9\x7f-\xff]*$');
        if (!codeRegx.test(this.currentBo.code)) {
            result.illegal = true;
            result.message = '业务对象的编号只能包含字母,数字,并且不可以用数字开头!';
            return result;
        }
        /** @type {?} */
        var nameRegx = new RegExp('[`~!@#$^&*()=|{}\':;\',\\[\\].<>/?~！@#￥……&*（）——|{}【】‘；：”“\'。，、？]');
        if (nameRegx.test(this.currentBo.name)) {
            result.illegal = true;
            result.message = '业务对象的名称包含非法字符，请重新输入!';
            return result;
        }
        return result;
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.repeatValidator = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var result = new ValidatorResult();
        if (!this.bosInSU || this.bosInSU.length <= 0) {
            return result;
        }
        /** @type {?} */
        var repeatCodeBO = this.bosInSU.find((/**
         * @param {?} bo
         * @return {?}
         */
        function (bo) { return bo.code.toLowerCase() === _this.currentBo.code.toLowerCase(); }));
        if (repeatCodeBO) {
            result.illegal = true;
            result.message = '同一SU下业务对象/业务对象分组/业务对象与业务对象分组编号不能重复，业务对象编号为：' + this.currentBo.code;
            return result;
        }
        return result;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    BoCreatorModalComponent.prototype.boGroupSelectChanged = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.parentBo = $event.data;
    };
    /**
     * @private
     * @return {?}
     */
    BoCreatorModalComponent.prototype.getParentBo = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.parentId) {
            this.boDtService.getBusinessObject(this.parentId).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                _this.parentBo = data;
                _this.suCode = data.code;
            }));
        }
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.clearHandler = /**
     * @return {?}
     */
    function () {
        this.parentBo = null;
    };
    BoCreatorModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-bo-creator-modal',
                    template: "<div class=\"cloumn\" style=\"margin-top: 28px;\">\n\n  <div class=\"suInfo row\">\n    <label class=\"col-form-label ps-col-form-label\">\n      <span class=\"farris-label-info text-danger\">*</span>\n      <span class=\"farris-label-text\">\u6240\u5C5E\u5FAE\u670D\u52A1</span>\n    </label>\n    <div class=\"col-{{inputLength}}\">\n      <input-group [(ngModel)]=\"suCode\" #boCode=\"ngModel\" [readonly]=\"true\"></input-group>\n    </div>\n  </div>\n\n  <div class=\"row suInfo\">\n    <label class=\"col-form-label ps-col-form-label\">\n      <span class=\"farris-label-text\">\u4E1A\u52A1\u5BF9\u8C61\u5206\u7EC4</span>\n    </label>\n    <div class=\"col-{{inputLength}}\">\n      <farris-combo-list #boGroupSelector [idField]=\"'id'\" [textField]=\"'name'\" [data]=\"items\"\n        (selectChange)=\"boGroupSelectChanged($event)\" (clear)=\"clearHandler()\" [panelHeight]=\"280\" [editable]=\"false\"></farris-combo-list>\n    </div>\n  </div>\n\n  <div class=\"row suInfo\">\n    <label class=\"col-form-label ps-col-form-label\">\n      <span class=\"farris-label-info text-danger\">*</span>\n      <span class=\"farris-label-text\">\u7F16\u53F7</span>\n    </label>\n    <div class=\"col-{{inputLength}}\">\n      <input-group [(ngModel)]=\"currentBo.code\" #boCode=\"ngModel\" required></input-group>\n    </div>\n    <!-- <div [hidden]=\"boCode.valid || boCode.pristine\" class=\"alert-danger\">\n      \u7F16\u53F7\u4E0D\u80FD\u4E3A\u7A7A\n    </div> -->\n  </div>\n\n  <div class=\"row suInfo\">\n    <label class=\"col-form-label ps-col-form-label\">\n      <span class=\"farris-label-info text-danger\">*</span>\n      <span class=\"farris-label-text\">\u540D\u79F0</span>\n    </label>\n    <div class=\"col-{{inputLength}}\">\n      <input-group [(ngModel)]=\"currentBo.name\" #boName=\"ngModel\" required></input-group>\n    </div>\n    <!-- <div [hidden]=\"boName.valid || boName.pristine\" class=\"alert-danger\">\n      \u540D\u79F0\u4E0D\u80FD\u4E3A\u7A7A\n    </div> -->\n  </div>\n\n</div>",
                    styles: [".sucreator-wrapper{background:rgba(255,255,255,.8);width:95%;margin:40px auto 0;max-width:480px;padding:24px 18px;border-radius:8px;position:relative;z-index:100}.suInfo{margin:0 0 16px 20px}.sucreator-wrapper .col-form-label,.sucreator-wrapper .form-control{padding-top:3px;padding-bottom:3px}.sucreator-wrapper .col-form-label{width:70px;margin-right:12px;text-align:right;flex-shrink:0}.bocreator{background:#fff;box-shadow:0 2px 16px 0 rgba(0,0,0,.1);border-radius:4px}.boCreatorPanel{background-color:#fff}div .ps-col-form-label{width:94px;text-align:right}"]
                }] }
    ];
    /** @nocollapse */
    BoCreatorModalComponent.ctorParameters = function () { return [
        { type: BoCreatorUtils },
        { type: BusinessObjectDtService },
        { type: NotifyService }
    ]; };
    BoCreatorModalComponent.propDecorators = {
        _parentId: [{ type: Input }],
        boConfirmEmitter: [{ type: Output }],
        boGroupSelector: [{ type: ViewChild, args: ["boGroupSelector",] }]
    };
    return BoCreatorModalComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/views/bo-creator/bo-creator.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BOCreatorComponent = /** @class */ (function () {
    function BOCreatorComponent() {
        this.createdAfterOk = new EventEmitter();
        this.modalClosed = new EventEmitter();
    }
    /**
     * @return {?}
     */
    BOCreatorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} $event
     * @return {?}
     */
    BOCreatorComponent.prototype.appConfirmEmitterHandler = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var bo = $event;
        this.createdAfterOk.emit(bo);
        this.modalClosed.emit();
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    BOCreatorComponent.prototype.suConfirmEmitterHandler = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var bo = $event;
        this.createdAfterOk.emit(bo);
        this.modalClosed.emit();
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    BOCreatorComponent.prototype.boConfirmEmitterHandler = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var bo = $event;
        this.createdAfterOk.emit(bo);
        this.modalClosed.emit();
    };
    /**
     * @return {?}
     */
    BOCreatorComponent.prototype.btnOkClick = /**
     * @return {?}
     */
    function () {
        switch (this.type) {
            case 'app':
                this.appCmp.create();
                return;
            case 'su':
                this.suCmp.create();
                return;
            case 'bo':
                this.boCmp.create();
                return;
            default:
                return;
        }
    };
    /**
     * @return {?}
     */
    BOCreatorComponent.prototype.hideModal = /**
     * @return {?}
     */
    function () {
        this.modalClosed.emit();
    };
    BOCreatorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-bo-creator',
                    template: "<div class=\"bo-creator farris-main-area flex-column farris-overflow-hidden\">\r\n\r\n  <div *ngIf=\"type==='app'\" class=\"modal-body boCreatorPanel\">\r\n    <app-app-creator-modal #appModal (appConfirmEmitter)=\"appConfirmEmitterHandler($event)\"></app-app-creator-modal>\r\n  </div>\r\n\r\n  <div *ngIf=\"type==='su'\" class=\"modal-body boCreatorPanel\">\r\n    <app-su-creator-modal #suModal (suConfirmEmitter)=\"suConfirmEmitterHandler($event)\" [(parentId)]=\"parentId\">\r\n    </app-su-creator-modal>\r\n  </div>\r\n\r\n  <div *ngIf=\"type ==='bo'\" class=\"modal-body boCreatorPanel\">\r\n    <app-bo-creator-modal #boModal (boConfirmEmitter)=\"boConfirmEmitterHandler($event)\" [(_parentId)]=\"parentId\">\r\n    </app-bo-creator-modal>\r\n  </div>\r\n\r\n  <div class=\"modal-footer btns\" #btns>\r\n    <button type=\"button\" class=\"btn btn-default btn-primary\" (click)=\"btnOkClick()\">\u786E\u5B9A</button>\r\n    <button type=\"button\" class=\"btn btn-default\" (click)=\"hideModal()\">\u53D6\u6D88</button>\r\n  </div>\r\n</div>",
                    styles: [".sucreator-wrapper{background:rgba(255,255,255,.8);width:95%;margin:40px auto 0;max-width:480px;padding:24px 18px;border-radius:8px;position:relative;z-index:100}.suInfo{margin:0 0 16px 20px}.sucreator-wrapper .col-form-label,.sucreator-wrapper .form-control{padding-top:3px;padding-bottom:3px}.sucreator-wrapper .col-form-label{width:70px;margin-right:12px;text-align:right;flex-shrink:0}.bocreator{background:#fff;box-shadow:0 2px 16px 0 rgba(0,0,0,.1);border-radius:4px}.boCreatorPanel{background-color:#fff}"]
                }] }
    ];
    /** @nocollapse */
    BOCreatorComponent.ctorParameters = function () { return []; };
    BOCreatorComponent.propDecorators = {
        createdAfterOk: [{ type: Output }],
        modalClosed: [{ type: Output }],
        parentId: [{ type: Input }],
        type: [{ type: Input }],
        appCmp: [{ type: ViewChild, args: ["appModal",] }],
        suCmp: [{ type: ViewChild, args: ["suModal",] }],
        boCmp: [{ type: ViewChild, args: ["boModal",] }]
    };
    return BOCreatorComponent;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bo-creator-package.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BOCreatorPackageModule = /** @class */ (function () {
    function BOCreatorPackageModule() {
    }
    BOCreatorPackageModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        BOCreatorComponent,
                        AppCreatorModalComponent,
                        SuCreatorModalComponent,
                        BoCreatorModalComponent
                    ],
                    imports: [
                        CommonModule,
                        InputGroupModule,
                        FormsModule,
                        InputsModule,
                        NotifyModule.forRoot(),
                        ComboListModule,
                        BoDtServiceModule
                    ],
                    entryComponents: [BOCreatorComponent],
                    exports: [BOCreatorComponent, AppCreatorModalComponent],
                    providers: [BusinessObjectDtService, BoCreatorUtils]
                },] }
    ];
    return BOCreatorPackageModule;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: gsp-lcm-bo-creator.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { BOCreatorPackageModule, BOCreatorComponent, AppCreatorModalComponent as ɵa, BoCreatorModalComponent as ɵd, SuCreatorModalComponent as ɵc, BoCreatorUtils as ɵb };
