// @farris/mobile-command-services/esm2015/lib/data-services/base-data.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/data-services/base-data.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EntityService } from '../entity-services/index';
import { LoadingService, ToastService, NotifyService, DialogService } from '../ui-services/index';
import { ExceptionService } from '../exception-services/index';
/**
 * 数据服务基类
 */
class BaseDataService {
    /**
     * @param {?} viewModelContext
     */
    constructor(viewModelContext) {
        this.viewModelContext = viewModelContext;
        this.init();
    }
    /**
     * @private
     * @return {?}
     */
    init() {
        this.initServices();
    }
    /**
     * @private
     * @return {?}
     */
    initServices() {
        this.loadingService = this.getService(LoadingService);
        this.toastService = this.getService(ToastService);
        this.notifyService = this.getService(NotifyService);
        this.dialogService = this.getService(DialogService);
        this.exceptionService = this.getService(ExceptionService);
        this.entityService = this.getService(EntityService);
        this.befRepository = (/** @type {?} */ (this.viewModelContext.repository));
    }
    /**
     * 获取服务实例
     * @template T
     * @param {?} token
     * @return {?}
     */
    getService(token) {
        /** @type {?} */
        const injector = this.viewModelContext.injector;
        return injector.get(token);
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.loadingService;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.toastService;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.notifyService;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.dialogService;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.exceptionService;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.befRepository;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.entityService;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.viewModelContext;
}
export { BaseDataService };
