// @gsp-bef/gsp-cm-metadata/esm5/lib/common-data-type/src/Element/GSPElementObjectType.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 字段类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
var GSPElementObjectType = {
    /**
     * 未设置
     */
    None: 'None',
    /**
     * 关联
     */
    Association: 'Association',
    /**
     * 枚举
     */
    Enum: 'Enum',
    /**
     * 动态属性
     */
    DynamicProp: 'DynamicProp',
};
export { GSPElementObjectType };
/** @type {?} */
export var GSPElementObjectTypes = [
    { index: 0, text: '无', value: GSPElementObjectType.None },
    { index: 1, text: '关联', value: GSPElementObjectType.Association },
    { index: 2, text: '枚举', value: GSPElementObjectType.Enum },
    { index: 3, text: '动态属性', value: GSPElementObjectType.DynamicProp },
];
var GetElementObjTypesUtil = /** @class */ (function () {
    function GetElementObjTypesUtil() {
    }
    /**
     * @return {?}
     */
    GetElementObjTypesUtil.GSPElementObjectTypesWithouDynamic = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var types = new Array();
        Object.assign(types, GSPElementObjectTypes);
        types = types.slice(0, 3);
        return types;
    };
    return GetElementObjTypesUtil;
}());
export { GetElementObjTypesUtil };
