// @gsp-bef/gsp-cm-metadata/esm5/lib/common-data-type/src/Element/GSPElementDataType.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 字段数据类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
var GSPElementDataType = {
    /**
     * 文本
     */
    String: 'String',
    /**
     * 备注
     */
    Text: 'Text',
    /**
     * 整数
     */
    Integer: 'Integer',
    /**
     * 浮点数
     */
    Decimal: 'Decimal',
    /**
     * 布尔型
     */
    Boolean: 'Boolean',
    /**
     * 日期型
     */
    Date: 'Date',
    /**
     * 日期时间型
     */
    DateTime: 'DateTime',
    /**
     * 二进制
     */
    Binary: 'Binary',
};
export { GSPElementDataType };
/** @type {?} */
export var GSPElementDataTypes = [
    {
        index: 0, text: '文本', value: GSPElementDataType.String, canChangeLength: true,
        canChangePresicion: false, length: 36, precision: 0
    },
    {
        index: 1, text: '备注', value: GSPElementDataType.Text, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    },
    {
        index: 2, text: '整数', value: GSPElementDataType.Integer, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    },
    {
        index: 3, text: '浮点数字', value: GSPElementDataType.Decimal, canChangeLength: true,
        canChangePresicion: true, length: 18, precision: 2
    },
    {
        index: 4, text: '布尔型', value: GSPElementDataType.Boolean, canChangeLength: false,
        canChangePresicion: false, length: 1, precision: 0
    },
    {
        index: 5, text: '日期', value: GSPElementDataType.Date, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    },
    {
        index: 6, text: '日期时间', value: GSPElementDataType.DateTime, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    },
    {
        index: 7, text: '二进制', value: GSPElementDataType.Binary, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    }, {
        index: 8, text: '业务字段', value: 'UnifiedDataType', canChangeLength: false,
        canChangePresicion: false, length: 36, precision: 0
    }
];
var GetDataTypesUtil = /** @class */ (function () {
    function GetDataTypesUtil() {
    }
    /**
     * @return {?}
     */
    GetDataTypesUtil.GSPElementDataTypesWithouUdt = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var types = new Array();
        Object.assign(types, GSPElementDataTypes);
        types = types.slice(0, 8);
        return types;
    };
    return GetDataTypesUtil;
}());
export { GetDataTypesUtil };
