// @farris/devkit/esm2015/lib/repository/sort_condition_manager.js
/**
 * 排序条件管理
 */
export class SortConditionManager {
    constructor() {
        this.sorts = new Map();
    }
    /**
     * 根据bingdingPath获取排序规则
     * @param bindingPath bindingPath
     * @param directionConverter converter
     */
    getConditionsByBindingPath(bindingPath, directionConverter) {
        let conditions = this.sorts.get(bindingPath) || [];
        if (conditions.length < 1) {
            return conditions;
        }
        if (typeof directionConverter === 'function') {
            conditions = conditions.map(condition => {
                const sortType = directionConverter(condition.SortType);
                return { SortField: condition.SortField, SortType: sortType };
            });
        }
        return conditions;
    }
    /**
     * 增加排序条件
     * @param bindingPath 绑定路径
     * @param field 排序字段
     * @param direction 排序方向 asc|desc
     */
    addCondition(bindingPath, field, direction) {
        // 取消排序
        if (!field || !direction) {
            this.sorts.delete(bindingPath);
            return;
        }
        const isBindingPathExist = this.sorts.has(bindingPath);
        const condition = { SortField: field, SortType: direction };
        if (!isBindingPathExist) {
            this.sorts.set(bindingPath, [condition]);
        }
        else {
            const bindingPathSortConditions = this.sorts.get(bindingPath) || [];
            // new condition will override the previous condition
            const conditionIndex = bindingPathSortConditions.findIndex(item => item.SortField === field);
            if (conditionIndex !== -1) {
                bindingPathSortConditions[conditionIndex] = condition;
            }
            else {
                bindingPathSortConditions.push(condition);
            }
        }
    }
    removeCondition(bindingPath, field) {
        throw new Error('not implement!');
    }
    /**
     * 批量增加排序条件
     * @param bindingPath bindingPath
     * @param fields fields , seprated by comma
     * @param directions directions , seprated by comma
     * @description set conditions and overwrite all conditions which has the same bindingPath
     */
    setConditions(bindingPath, fields, directions) {
        // 取消排序
        if (!fields || !directions) {
            this.sorts.delete(bindingPath);
            return;
        }
        const filedsArray = fields.split(',').filter(item => item);
        const directionsArray = directions.split(',').filter(item => item);
        if (filedsArray.length !== directionsArray.length) {
            throw new Error('arguments error,fields and direction are not match.');
        }
        const conditions = [];
        filedsArray.forEach((field, index) => {
            const condition = { SortField: field, SortType: directionsArray[index] };
            conditions.push(condition);
        });
        this.sorts.set(bindingPath, conditions);
    }
    /**
     * 清空所有排序条件
     */
    clear() {
        this.sorts.clear();
    }
}
