// @edp-aif/common-api/esm5/lib/entity/impl/AifConfigEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
import { Entity } from './Entity';
var AifConfigEntity = /** @class */ (function (_super) {
    tslib_1.__extends(AifConfigEntity, _super);
    function AifConfigEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    AifConfigEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.content = null;
        this.tag = null;
    };
    /**
     * @return {?}
     */
    AifConfigEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tag", this.tag);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AifConfigEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
        this.tag = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tag");
    };
    return AifConfigEntity;
}(Entity));
export { AifConfigEntity };
if (false) {
    /** @type {?} */
    AifConfigEntity.prototype.content;
    /** @type {?} */
    AifConfigEntity.prototype.tag;
}
