// @edp-aif/common-api/esm5/lib/entity/impl/AssociationEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Entity } from './Entity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
var AssociationEntity = /** @class */ (function (_super) {
    tslib_1.__extends(AssociationEntity, _super);
    function AssociationEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    AssociationEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.creationRuleId = null;
        this.creationRuleInstanceId = null;
        this.bizFlowchartId = null;
        this.creationProcessId = null;
        this.creationProcessInstanceId = null;
        this.srcBeMetadataId = null;
        this.tarBeMetadataId = null;
        this.tarBoId = null;
        this.srcBoId = null;
        this.srcAppCode = null;
        this.tarAppCode = null;
        this.srcSuCode = null;
        this.tarSuCode = null;
        //this.srcBeType = null;
        //this.tarBeType = null;
        this.relationName = null;
        this.srcRootTableName = null;
        this.srcRootDataId = null;
        this.srcParentTableName = null;
        this.srcParentDataId = null;
        this.srcTableName = null;
        this.srcDataId = null;
        this.tarRootTableName = null;
        this.tarRootDataId = null;
        this.tarParentTableName = null;
        this.tarParentDataId = null;
        this.tarTableName = null;
        this.tarDataId = null;
        //this.srcVoType = null;
        //this.tarVoType = null;
        //this.tarVoProxyType = null;
        this.srcVoMetadataId = null;
        this.tarVoMetadataId = null;
        this.tarVoProxyMetadataId = null;
        this.srcTableNamePath = null;
        this.srcDataIdPath = null;
        this.tarTableNamePath = null;
        this.tarDataIdPath = null;
        this.srcBeId = null;
        this.tarBeId = null;
        this.srcVoId = null;
        this.tarVoId = null;
        this.tarVoProxyId = null;
        this.srcBeMetadataRootId = null;
        this.srcBeMetadataParentId = null;
        this.srcBeMetadataIdPath = null;
        this.srcBeRootId = null;
        this.srcBeParentId = null;
        this.srcBeIdPath = null;
        this.srcVoMetadataRootId = null;
        this.srcVoMetadataParentId = null;
        this.srcVoMetadataIdPath = null;
        this.srcVoRootId = null;
        this.srcVoParentId = null;
        this.srcVoIdPath = null;
        this.tarBeMetadataRootId = null;
        this.tarBeMetadataParentId = null;
        this.tarBeMetadataIdPath = null;
        this.tarBeRootId = null;
        this.tarBeParentId = null;
        this.tarBeIdPath = null;
        this.tarVoMetadataRootId = null;
        this.tarVoMetadataParentId = null;
        this.tarVoMetadataIdPath = null;
        this.tarVoRootId = null;
        this.tarVoParentId = null;
        this.tarVoIdPath = null;
        this.tarVoProxyMetadataRootId = null;
        this.tarVoProxyMetadataParentId = null;
        this.tarVoProxyMetadataIdPath = null;
        this.tarVoProxyRootId = null;
        this.tarVoProxyParentId = null;
        this.tarVoProxyIdPath = null;
    };
    /**
     * @return {?}
     */
    AssociationEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessId", this.creationProcessId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessInstanceId", this.creationProcessInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcBeMetadataId", this.srcBeMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarBeMetadataId", this.tarBeMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarBoId", this.tarBoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcBoId", this.srcBoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcAppCode", this.srcAppCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarAppCode", this.tarAppCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcSuCode", this.srcSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarSuCode", this.tarSuCode);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "srcBeType", this.srcBeType);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "tarBeType", this.tarBeType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "relationName", this.relationName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcRootTableName", this.srcRootTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcRootDataId", this.srcRootDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcParentTableName", this.srcParentTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcParentDataId", this.srcParentDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcTableName", this.srcTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcDataId", this.srcDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarRootTableName", this.tarRootTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarRootDataId", this.tarRootDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarParentTableName", this.tarParentTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarParentDataId", this.tarParentDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarTableName", this.tarTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarDataId", this.tarDataId);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "srcVoType", this.srcVoType);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoType", this.tarVoType);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoProxyType", this.tarVoProxyType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcVoMetadataId", this.srcVoMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoMetadataId", this.tarVoMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoProxyMetadataId", this.tarVoProxyMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcTableNamePath", this.srcTableNamePath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcDataIdPath", this.srcDataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarTableNamePath", this.tarTableNamePath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarDataIdPath", this.tarDataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcBeId", this.srcBeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarBeId", this.tarBeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcVoId", this.srcVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoId", this.tarVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoProxyId", this.tarVoProxyId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcBeMetadataRootId", this.srcBeMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcBeMetadataParentId", this.srcBeMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcBeMetadataIdPath", this.srcBeMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcBeRootId", this.srcBeRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcBeParentId", this.srcBeParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcBeIdPath", this.srcBeIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcVoMetadataRootId", this.srcVoMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcVoMetadataParentId", this.srcVoMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcVoMetadataIdPath", this.srcVoMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcVoRootId", this.srcVoRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcVoParentId", this.srcVoParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcVoIdPath", this.srcVoIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarBeMetadataRootId", this.tarBeMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarBeMetadataParentId", this.tarBeMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarBeMetadataIdPath", this.tarBeMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarBeRootId", this.tarBeRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarBeParentId", this.tarBeParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarBeIdPath", this.tarBeIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoMetadataRootId", this.tarVoMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoMetadataParentId", this.tarVoMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoMetadataIdPath", this.tarVoMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoRootId", this.tarVoRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoParentId", this.tarVoParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoIdPath", this.tarVoIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoProxyMetadataRootId", this.tarVoProxyMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoProxyMetadataParentId", this.tarVoProxyMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoProxyMetadataIdPath", this.tarVoProxyMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoProxyRootId", this.tarVoProxyRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoProxyParentId", this.tarVoProxyParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarVoProxyIdPath", this.tarVoProxyIdPath);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AssociationEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.creationRuleInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.creationProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessId");
        this.creationProcessInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessInstanceId");
        this.srcBeMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcBeMetadataId");
        this.tarBeMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarBeMetadataId");
        this.tarBoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarBoId");
        this.srcBoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcBoId");
        this.srcAppCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcAppCode");
        this.tarAppCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarAppCode");
        this.srcSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcSuCode");
        this.tarSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarSuCode");
        //this.srcBeType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcBeType");
        //this.tarBeType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarBeType");
        this.relationName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "relationName");
        this.srcRootTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcRootTableName");
        this.srcRootDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcRootDataId");
        this.srcParentTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcParentTableName");
        this.srcParentDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcParentDataId");
        this.srcTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcTableName");
        this.srcDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcDataId");
        this.tarRootTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarRootTableName");
        this.tarRootDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarRootDataId");
        this.tarParentTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarParentTableName");
        this.tarParentDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarParentDataId");
        this.tarTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarTableName");
        this.tarDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarDataId");
        //this.srcVoType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcVoType");
        //this.tarVoType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoType");
        //this.tarVoProxyType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoProxyType");
        this.srcVoMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcVoMetadataId");
        this.tarVoMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoMetadataId");
        this.tarVoProxyMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoProxyMetadataId");
        this.srcTableNamePath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcTableNamePath");
        this.srcDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcDataIdPath");
        this.tarTableNamePath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarTableNamePath");
        this.tarDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarDataIdPath");
        this.srcBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcBeId");
        this.tarBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarBeId");
        this.srcVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcVoId");
        this.tarVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoId");
        this.tarVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoProxyId");
        this.srcBeMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcBeMetadataRootId");
        this.srcBeMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcBeMetadataParentId");
        this.srcBeMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcBeMetadataIdPath");
        this.srcBeRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcBeRootId");
        this.srcBeParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcBeParentId");
        this.srcBeIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcBeIdPath");
        this.srcVoMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcVoMetadataRootId");
        this.srcVoMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcVoMetadataParentId");
        this.srcVoMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcVoMetadataIdPath");
        this.srcVoRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcVoRootId");
        this.srcVoParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcVoParentId");
        this.srcVoIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcVoIdPath");
        this.tarBeMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarBeMetadataRootId");
        this.tarBeMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarBeMetadataParentId");
        this.tarBeMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarBeMetadataIdPath");
        this.tarBeRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarBeRootId");
        this.tarBeParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarBeParentId");
        this.tarBeIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarBeIdPath");
        this.tarVoMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoMetadataRootId");
        this.tarVoMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoMetadataParentId");
        this.tarVoMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoMetadataIdPath");
        this.tarVoRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoRootId");
        this.tarVoParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoParentId");
        this.tarVoIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoIdPath");
        this.tarVoProxyMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoProxyMetadataRootId");
        this.tarVoProxyMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoProxyMetadataParentId");
        this.tarVoProxyMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoProxyMetadataIdPath");
        this.tarVoProxyRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoProxyRootId");
        this.tarVoProxyParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoProxyParentId");
        this.tarVoProxyIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarVoProxyIdPath");
    };
    return AssociationEntity;
}(Entity));
export { AssociationEntity };
if (false) {
    /** @type {?} */
    AssociationEntity.prototype.creationRuleId;
    /** @type {?} */
    AssociationEntity.prototype.creationRuleInstanceId;
    /** @type {?} */
    AssociationEntity.prototype.bizFlowchartId;
    /** @type {?} */
    AssociationEntity.prototype.creationProcessId;
    /** @type {?} */
    AssociationEntity.prototype.creationProcessInstanceId;
    /** @type {?} */
    AssociationEntity.prototype.srcBoId;
    /** @type {?} */
    AssociationEntity.prototype.tarBoId;
    /** @type {?} */
    AssociationEntity.prototype.srcAppCode;
    /** @type {?} */
    AssociationEntity.prototype.tarAppCode;
    /** @type {?} */
    AssociationEntity.prototype.srcSuCode;
    /** @type {?} */
    AssociationEntity.prototype.tarSuCode;
    /** @type {?} */
    AssociationEntity.prototype.relationName;
    /** @type {?} */
    AssociationEntity.prototype.srcRootTableName;
    /** @type {?} */
    AssociationEntity.prototype.srcRootDataId;
    /** @type {?} */
    AssociationEntity.prototype.srcParentTableName;
    /** @type {?} */
    AssociationEntity.prototype.srcParentDataId;
    /** @type {?} */
    AssociationEntity.prototype.tarRootTableName;
    /** @type {?} */
    AssociationEntity.prototype.tarRootDataId;
    /** @type {?} */
    AssociationEntity.prototype.tarParentTableName;
    /** @type {?} */
    AssociationEntity.prototype.tarParentDataId;
    /** @type {?} */
    AssociationEntity.prototype.srcVoMetadataId;
    /** @type {?} */
    AssociationEntity.prototype.tarVoMetadataId;
    /** @type {?} */
    AssociationEntity.prototype.tarVoProxyMetadataId;
    /** @type {?} */
    AssociationEntity.prototype.srcTableNamePath;
    /** @type {?} */
    AssociationEntity.prototype.srcDataIdPath;
    /** @type {?} */
    AssociationEntity.prototype.tarTableNamePath;
    /** @type {?} */
    AssociationEntity.prototype.tarDataIdPath;
    /** @type {?} */
    AssociationEntity.prototype.srcBeId;
    /** @type {?} */
    AssociationEntity.prototype.tarBeId;
    /** @type {?} */
    AssociationEntity.prototype.srcVoId;
    /** @type {?} */
    AssociationEntity.prototype.tarVoId;
    /** @type {?} */
    AssociationEntity.prototype.tarVoProxyId;
    /** @type {?} */
    AssociationEntity.prototype.srcBeMetadataRootId;
    /** @type {?} */
    AssociationEntity.prototype.srcBeMetadataParentId;
    /** @type {?} */
    AssociationEntity.prototype.srcBeMetadataIdPath;
    /** @type {?} */
    AssociationEntity.prototype.srcBeRootId;
    /** @type {?} */
    AssociationEntity.prototype.srcBeParentId;
    /** @type {?} */
    AssociationEntity.prototype.srcBeIdPath;
    /** @type {?} */
    AssociationEntity.prototype.srcVoMetadataRootId;
    /** @type {?} */
    AssociationEntity.prototype.srcVoMetadataParentId;
    /** @type {?} */
    AssociationEntity.prototype.srcVoMetadataIdPath;
    /** @type {?} */
    AssociationEntity.prototype.srcVoRootId;
    /** @type {?} */
    AssociationEntity.prototype.srcVoParentId;
    /** @type {?} */
    AssociationEntity.prototype.srcVoIdPath;
    /** @type {?} */
    AssociationEntity.prototype.tarBeMetadataRootId;
    /** @type {?} */
    AssociationEntity.prototype.tarBeMetadataParentId;
    /** @type {?} */
    AssociationEntity.prototype.tarBeMetadataIdPath;
    /** @type {?} */
    AssociationEntity.prototype.tarBeRootId;
    /** @type {?} */
    AssociationEntity.prototype.tarBeParentId;
    /** @type {?} */
    AssociationEntity.prototype.tarBeIdPath;
    /** @type {?} */
    AssociationEntity.prototype.tarVoMetadataRootId;
    /** @type {?} */
    AssociationEntity.prototype.tarVoMetadataParentId;
    /** @type {?} */
    AssociationEntity.prototype.tarVoMetadataIdPath;
    /** @type {?} */
    AssociationEntity.prototype.tarVoRootId;
    /** @type {?} */
    AssociationEntity.prototype.tarVoParentId;
    /** @type {?} */
    AssociationEntity.prototype.tarVoIdPath;
    /** @type {?} */
    AssociationEntity.prototype.tarVoProxyMetadataRootId;
    /** @type {?} */
    AssociationEntity.prototype.tarVoProxyMetadataParentId;
    /** @type {?} */
    AssociationEntity.prototype.tarVoProxyMetadataIdPath;
    /** @type {?} */
    AssociationEntity.prototype.tarVoProxyRootId;
    /** @type {?} */
    AssociationEntity.prototype.tarVoProxyParentId;
    /** @type {?} */
    AssociationEntity.prototype.tarVoProxyIdPath;
    /** @type {?} */
    AssociationEntity.prototype.srcBeMetadataId;
    /** @type {?} */
    AssociationEntity.prototype.tarBeMetadataId;
    /** @type {?} */
    AssociationEntity.prototype.srcDataId;
    /** @type {?} */
    AssociationEntity.prototype.srcTableName;
    /** @type {?} */
    AssociationEntity.prototype.tarDataId;
    /** @type {?} */
    AssociationEntity.prototype.tarTableName;
}
