// @edp-aif/common-api/esm5/lib/entity/impl/BizActionEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { MetadataEntity } from './MetadataEntity';
import { BizActionType } from '../BizActionType';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * @record
 */
export function BizActionEntityConstructor() { }
/**
 * 业务动作实体类。
 */
var /**
 * 业务动作实体类。
 */
BizActionEntity = /** @class */ (function (_super) {
    tslib_1.__extends(BizActionEntity, _super);
    function BizActionEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    BizActionEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.bizActionType = BizActionType.DEFAULT;
        this.beMetadataId = null;
        //this.beType = null;
        this.bizActionAgentMetadataId = null;
        this.bizActionAgentComponentId = null;
        this.voMetadataId = null;
        //this.voType = null;
        this.beId = null;
        this.voId = null;
        this.bizActionAgentClassTypeName = null;
        this.beMetadataRootId = null;
        this.beMetadataParentId = null;
        this.beMetadataIdPath = null;
        this.beRootId = null;
        this.beParentId = null;
        this.beIdPath = null;
        this.voMetadataRootId = null;
        this.voMetadataParentId = null;
        this.voMetadataIdPath = null;
        this.voRootId = null;
        this.voParentId = null;
        this.voIdPath = null;
    };
    /**
     * @return {?}
     */
    BizActionEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionType", this.bizActionType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beMetadataId", this.beMetadataId);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "beType", this.beType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionAgentMetadataId", this.bizActionAgentMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionAgentComponentId", this.bizActionAgentComponentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voMetadataId", this.voMetadataId);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "voType", this.voType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beId", this.beId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionAgentClassTypeName", this.bizActionAgentClassTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beMetadataRootId", this.beMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beMetadataParentId", this.beMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beMetadataIdPath", this.beMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beRootId", this.beRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beParentId", this.beParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beIdPath", this.beIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voMetadataRootId", this.voMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voMetadataParentId", this.voMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voMetadataIdPath", this.voMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voRootId", this.voRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voParentId", this.voParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voIdPath", this.voIdPath);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BizActionEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.bizActionType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionType");
        this.beMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beMetadataId");
        //this.beType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beType");
        this.bizActionAgentMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionAgentMetadataId");
        this.bizActionAgentComponentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionAgentComponentId");
        this.voMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voMetadataId");
        //this.voType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voType");
        this.beId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beId");
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
        this.bizActionAgentClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionAgentClassTypeName");
        this.beMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beMetadataRootId");
        this.beMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beMetadataParentId");
        this.beMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beMetadataIdPath");
        this.beRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beRootId");
        this.beParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beParentId");
        this.beIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beIdPath");
        this.voMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voMetadataRootId");
        this.voMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voMetadataParentId");
        this.voMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voMetadataIdPath");
        this.voRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voRootId");
        this.voParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voParentId");
        this.voIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voIdPath");
    };
    return BizActionEntity;
}(MetadataEntity));
/**
 * 业务动作实体类。
 */
export { BizActionEntity };
if (false) {
    /** @type {?} */
    BizActionEntity.prototype.beMetadataId;
    /** @type {?} */
    BizActionEntity.prototype.bizActionAgentMetadataId;
    /** @type {?} */
    BizActionEntity.prototype.voMetadataId;
    /** @type {?} */
    BizActionEntity.prototype.beId;
    /** @type {?} */
    BizActionEntity.prototype.beMetadataRootId;
    /** @type {?} */
    BizActionEntity.prototype.beMetadataParentId;
    /** @type {?} */
    BizActionEntity.prototype.beMetadataIdPath;
    /** @type {?} */
    BizActionEntity.prototype.beRootId;
    /** @type {?} */
    BizActionEntity.prototype.beParentId;
    /** @type {?} */
    BizActionEntity.prototype.beIdPath;
    /** @type {?} */
    BizActionEntity.prototype.voMetadataRootId;
    /** @type {?} */
    BizActionEntity.prototype.voMetadataParentId;
    /** @type {?} */
    BizActionEntity.prototype.voMetadataIdPath;
    /** @type {?} */
    BizActionEntity.prototype.voRootId;
    /** @type {?} */
    BizActionEntity.prototype.voParentId;
    /** @type {?} */
    BizActionEntity.prototype.voIdPath;
    /** @type {?} */
    BizActionEntity.prototype.bizActionType;
    /** @type {?} */
    BizActionEntity.prototype.voId;
    /** @type {?} */
    BizActionEntity.prototype.bizActionAgentComponentId;
    /** @type {?} */
    BizActionEntity.prototype.bizActionAgentClassTypeName;
}
