// @edp-aif/common-api/esm5/lib/entity/impl/BizEventEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { MetadataEntity } from './MetadataEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * 业务事件实体类。
 */
var /**
 * 业务事件实体类。
 */
BizEventEntity = /** @class */ (function (_super) {
    tslib_1.__extends(BizEventEntity, _super);
    function BizEventEntity() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isEnabled = true;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    BizEventEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.isEnabled = true;
        this.beEventRouterMetadataId = null;
        this.beEventRouterComponentId = null;
        this.bizEventAgentMetadataId = null;
        this.bizEventAgentComponentId = null;
        this.beMetadataId = null;
        this.beId = null;
        //this.beType = null;
        this.voMetadataId = null;
        this.voId = null;
        //this.voType = null;
        this.canDoExpressionId = null;
        this.bizFlowchartIdExpressionId = null;
        this.beEventRouterClassTypeName = null;
        this.beMetadataRootId = null;
        this.beMetadataParentId = null;
        this.beMetadataIdPath = null;
        this.beRootId = null;
        this.beParentId = null;
        this.beIdPath = null;
        this.voMetadataRootId = null;
        this.voMetadataParentId = null;
        this.voMetadataIdPath = null;
        this.voRootId = null;
        this.voParentId = null;
        this.voIdPath = null;
    };
    /**
     * @return {?}
     */
    BizEventEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beEventRouterMetadataId", this.beEventRouterMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beEventRouterComponentId", this.beEventRouterComponentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventAgentMetadataId", this.bizEventAgentMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventAgentComponentId", this.bizEventAgentComponentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beMetadataId", this.beMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beId", this.beId);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "beType", this.beType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voMetadataId", this.voMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "voType", this.voType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "canDoExpressionId", this.canDoExpressionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartIdExpressionId", this.bizFlowchartIdExpressionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beEventRouterClassTypeName", this.beEventRouterClassTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beMetadataRootId", this.beMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beMetadataParentId", this.beMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beMetadataIdPath", this.beMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beRootId", this.beRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beParentId", this.beParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beIdPath", this.beIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voMetadataRootId", this.voMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voMetadataParentId", this.voMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voMetadataIdPath", this.voMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voRootId", this.voRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voParentId", this.voParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voIdPath", this.voIdPath);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BizEventEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.beEventRouterMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beEventRouterMetadataId");
        this.beEventRouterComponentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beEventRouterComponentId");
        this.bizEventAgentMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventAgentMetadataId");
        this.bizEventAgentComponentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventAgentComponentId");
        this.beMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beMetadataId");
        //this.beType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beType");
        this.beId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beId");
        this.voMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voMetadataId");
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
        //this.voType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voType");
        this.canDoExpressionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "canDoExpressionId");
        this.bizFlowchartIdExpressionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartIdExpressionId");
        this.beEventRouterClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beEventRouterClassTypeName");
        this.beMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beMetadataRootId");
        this.beMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beMetadataParentId");
        this.beMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beMetadataIdPath");
        this.beRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beRootId");
        this.beParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beParentId");
        this.beIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beIdPath");
        this.voMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voMetadataRootId");
        this.voMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voMetadataParentId");
        this.voMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voMetadataIdPath");
        this.voRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voRootId");
        this.voParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voParentId");
        this.voIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voIdPath");
    };
    return BizEventEntity;
}(MetadataEntity));
/**
 * 业务事件实体类。
 */
export { BizEventEntity };
if (false) {
    /** @type {?} */
    BizEventEntity.prototype.isEnabled;
    /** @type {?} */
    BizEventEntity.prototype.beEventRouterMetadataId;
    /** @type {?} */
    BizEventEntity.prototype.beEventRouterComponentId;
    /** @type {?} */
    BizEventEntity.prototype.bizEventAgentMetadataId;
    /** @type {?} */
    BizEventEntity.prototype.bizEventAgentComponentId;
    /** @type {?} */
    BizEventEntity.prototype.beMetadataId;
    /** @type {?} */
    BizEventEntity.prototype.beId;
    /** @type {?} */
    BizEventEntity.prototype.voMetadataId;
    /** @type {?} */
    BizEventEntity.prototype.voId;
    /** @type {?} */
    BizEventEntity.prototype.canDoExpressionId;
    /** @type {?} */
    BizEventEntity.prototype.bizFlowchartIdExpressionId;
    /** @type {?} */
    BizEventEntity.prototype.beEventRouterClassTypeName;
    /** @type {?} */
    BizEventEntity.prototype.beMetadataRootId;
    /** @type {?} */
    BizEventEntity.prototype.beMetadataParentId;
    /** @type {?} */
    BizEventEntity.prototype.beMetadataIdPath;
    /** @type {?} */
    BizEventEntity.prototype.beRootId;
    /** @type {?} */
    BizEventEntity.prototype.beParentId;
    /** @type {?} */
    BizEventEntity.prototype.beIdPath;
    /** @type {?} */
    BizEventEntity.prototype.voMetadataRootId;
    /** @type {?} */
    BizEventEntity.prototype.voMetadataParentId;
    /** @type {?} */
    BizEventEntity.prototype.voMetadataIdPath;
    /** @type {?} */
    BizEventEntity.prototype.voRootId;
    /** @type {?} */
    BizEventEntity.prototype.voParentId;
    /** @type {?} */
    BizEventEntity.prototype.voIdPath;
}
