// @edp-aif/common-api/esm5/lib/entity/impl/BizFlowchartRefEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Entity } from './Entity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
var BizFlowchartRefEntity = /** @class */ (function (_super) {
    tslib_1.__extends(BizFlowchartRefEntity, _super);
    function BizFlowchartRefEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    BizFlowchartRefEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.bizFlowchartId = null;
        this.refEntityId = null;
    };
    /**
     * @return {?}
     */
    BizFlowchartRefEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "refEntityId", this.refEntityId);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BizFlowchartRefEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.refEntityId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "refEntityId");
    };
    return BizFlowchartRefEntity;
}(Entity));
export { BizFlowchartRefEntity };
if (false) {
    /** @type {?} */
    BizFlowchartRefEntity.prototype.bizFlowchartId;
    /** @type {?} */
    BizFlowchartRefEntity.prototype.refEntityId;
}
