// @edp-aif/common-api/esm5/lib/entity/impl/BizFlowchartRefMetadataEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BizFlowchartRefEntity } from './BizFlowchartRefEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
var BizFlowchartRefMetadataEntity = /** @class */ (function (_super) {
    tslib_1.__extends(BizFlowchartRefMetadataEntity, _super);
    function BizFlowchartRefMetadataEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    BizFlowchartRefMetadataEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.refAppCode = null;
        this.refSuCode = null;
        this.refBoId = null;
        this.bizFlowchartAppCode = null;
        this.bizFlowchartSuCode = null;
        this.bizFlowchartBoId = null;
    };
    /**
     * @return {?}
     */
    BizFlowchartRefMetadataEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "refAppCode", this.refAppCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "refSuCode", this.refSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "refBoId", this.refBoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartAppCode", this.bizFlowchartAppCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartSuCode", this.bizFlowchartSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartBoId", this.bizFlowchartBoId);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BizFlowchartRefMetadataEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.refAppCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "refAppCode");
        this.refSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "refSuCode");
        this.refBoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "refBoId");
        this.bizFlowchartAppCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartAppCode");
        this.bizFlowchartSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartSuCode");
        this.bizFlowchartBoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartBoId");
    };
    return BizFlowchartRefMetadataEntity;
}(BizFlowchartRefEntity));
export { BizFlowchartRefMetadataEntity };
if (false) {
    /** @type {?} */
    BizFlowchartRefMetadataEntity.prototype.refAppCode;
    /** @type {?} */
    BizFlowchartRefMetadataEntity.prototype.refSuCode;
    /** @type {?} */
    BizFlowchartRefMetadataEntity.prototype.refBoId;
    /** @type {?} */
    BizFlowchartRefMetadataEntity.prototype.bizFlowchartAppCode;
    /** @type {?} */
    BizFlowchartRefMetadataEntity.prototype.bizFlowchartSuCode;
    /** @type {?} */
    BizFlowchartRefMetadataEntity.prototype.bizFlowchartBoId;
}
