// @edp-aif/common-api/esm5/lib/entity/impl/BizProcessEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { P2PMetadataEntity } from './P2PMetadataEntity';
import { CallType } from '../CallType';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * 业务流程实体类。
 * @abstract
 */
var /**
 * 业务流程实体类。
 * @abstract
 */
BizProcessEntity = /** @class */ (function (_super) {
    tslib_1.__extends(BizProcessEntity, _super);
    function BizProcessEntity() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.callType = CallType.SYNC;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    BizProcessEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.bizEventId = null;
        this.bizActionId = null;
        this.isEnabled = false;
        this.callType = CallType.SYNC;
    };
    /**
     * @return {?}
     */
    BizProcessEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventId", this.bizEventId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionId", this.bizActionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "callType", this.callType);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BizProcessEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.bizEventId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventId");
        this.bizActionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionId");
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.callType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "callType");
    };
    return BizProcessEntity;
}(P2PMetadataEntity));
/**
 * 业务流程实体类。
 * @abstract
 */
export { BizProcessEntity };
if (false) {
    /** @type {?} */
    BizProcessEntity.prototype.bizEventId;
    /** @type {?} */
    BizProcessEntity.prototype.bizActionId;
    /** @type {?} */
    BizProcessEntity.prototype.isEnabled;
    /** @type {?} */
    BizProcessEntity.prototype.callType;
}
