// @edp-aif/common-api/esm5/lib/entity/impl/CreationRuleEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { MappingEntity } from './MappingEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * 生单规则实体类
 */
var /**
 * 生单规则实体类
 */
CreationRuleEntity = /** @class */ (function (_super) {
    tslib_1.__extends(CreationRuleEntity, _super);
    function CreationRuleEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    CreationRuleEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.isEnabled = false;
        this.formSettingsId = null;
    };
    /**
     * @return {?}
     */
    CreationRuleEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "formSettingsId", this.formSettingsId);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CreationRuleEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.formSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettingsId");
    };
    return CreationRuleEntity;
}(MappingEntity));
/**
 * 生单规则实体类
 */
export { CreationRuleEntity };
if (false) {
    /** @type {?} */
    CreationRuleEntity.prototype.isEnabled;
    /** @type {?} */
    CreationRuleEntity.prototype.formSettingsId;
}
