// @edp-aif/common-api/esm5/lib/entity/impl/CreationProcessEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BizProcessEntity } from './BizProcessEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * 生单流程实体类。
 */
var /**
 * 生单流程实体类。
 */
CreationProcessEntity = /** @class */ (function (_super) {
    tslib_1.__extends(CreationProcessEntity, _super);
    function CreationProcessEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    CreationProcessEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.creationRuleId = null;
        this.reverseWriteRuleId = null;
        this.reverseWriteRuleClassName = null;
    };
    /**
     * @return {?}
     */
    CreationProcessEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleId", this.reverseWriteRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleClassName", this.reverseWriteRuleClassName);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CreationProcessEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.reverseWriteRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleId");
        this.reverseWriteRuleClassName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleClassName");
    };
    return CreationProcessEntity;
}(BizProcessEntity));
/**
 * 生单流程实体类。
 */
export { CreationProcessEntity };
if (false) {
    /** @type {?} */
    CreationProcessEntity.prototype.creationRuleId;
    /** @type {?} */
    CreationProcessEntity.prototype.reverseWriteRuleId;
    /** @type {?} */
    CreationProcessEntity.prototype.reverseWriteRuleClassName;
}
