// @edp-aif/common-api/esm5/lib/entity/impl/GeneralTextEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Entity } from './Entity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
var GeneralTextEntity = /** @class */ (function (_super) {
    tslib_1.__extends(GeneralTextEntity, _super);
    function GeneralTextEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    GeneralTextEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.text = null;
        this.tag = null;
    };
    /**
     * @return {?}
     */
    GeneralTextEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "text", this.text);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tag", this.tag);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    GeneralTextEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.text = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "text");
        this.tag = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tag");
    };
    return GeneralTextEntity;
}(Entity));
export { GeneralTextEntity };
if (false) {
    /** @type {?} */
    GeneralTextEntity.prototype.text;
    /** @type {?} */
    GeneralTextEntity.prototype.tag;
}
