// @edp-aif/common-api/esm5/lib/entity/impl/MappingEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { P2PMetadataEntity } from './P2PMetadataEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * 映射规则实体类接口。
 * @abstract
 */
var /**
 * 映射规则实体类接口。
 * @abstract
 */
MappingEntity = /** @class */ (function (_super) {
    tslib_1.__extends(MappingEntity, _super);
    function MappingEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    MappingEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.mappingRuleMetadataId = null;
        this.mappingRuleId = null;
        this.mappingComponentMetadataId = null;
        this.mappingComponentId = null;
        this.mappingClassTypeName = null;
    };
    /**
     * @return {?}
     */
    MappingEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleMetadataId", this.mappingRuleMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleId", this.mappingRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingComponentMetadataId", this.mappingComponentMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingComponentId", this.mappingComponentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingClassTypeName", this.mappingClassTypeName);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    MappingEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.mappingRuleMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleMetadataId");
        this.mappingRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleId");
        this.mappingComponentMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingComponentMetadataId");
        this.mappingComponentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingComponentId");
        this.mappingClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingClassTypeName");
    };
    return MappingEntity;
}(P2PMetadataEntity));
/**
 * 映射规则实体类接口。
 * @abstract
 */
export { MappingEntity };
if (false) {
    /** @type {?} */
    MappingEntity.prototype.mappingRuleMetadataId;
    /** @type {?} */
    MappingEntity.prototype.mappingRuleId;
    /** @type {?} */
    MappingEntity.prototype.mappingComponentMetadataId;
    /** @type {?} */
    MappingEntity.prototype.mappingComponentId;
    /** @type {?} */
    MappingEntity.prototype.mappingClassTypeName;
}
