// @edp-aif/common-api/esm5/lib/entity/impl/MetadataEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Entity } from './Entity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * 元数据实体类接口。
 * @abstract
 */
var /**
 * 元数据实体类接口。
 * @abstract
 */
MetadataEntity = /** @class */ (function (_super) {
    tslib_1.__extends(MetadataEntity, _super);
    function MetadataEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    MetadataEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.appCode = null;
        this.suCode = null;
        this.boId = null;
        this.persistenceVersion = null;
        this.name = null;
        this.description = null;
    };
    /**
     * @return {?}
     */
    MetadataEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "appCode", this.appCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "boId", this.boId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "persistenceVersion", this.persistenceVersion);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "description", this.description);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    MetadataEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.appCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "appCode");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.boId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "boId");
        this.persistenceVersion = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "persistenceVersion");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.description = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "description");
    };
    return MetadataEntity;
}(Entity));
/**
 * 元数据实体类接口。
 * @abstract
 */
export { MetadataEntity };
if (false) {
    /** @type {?} */
    MetadataEntity.prototype.appCode;
    /** @type {?} */
    MetadataEntity.prototype.suCode;
    /** @type {?} */
    MetadataEntity.prototype.boId;
    /** @type {?} */
    MetadataEntity.prototype.persistenceVersion;
    /** @type {?} */
    MetadataEntity.prototype.name;
    /** @type {?} */
    MetadataEntity.prototype.description;
}
