// @edp-aif/common-api/esm5/lib/entity/impl/P2PMetadataEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { MetadataEntity } from './MetadataEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * 点到点元数据实体类。
 * @abstract
 */
var /**
 * 点到点元数据实体类。
 * @abstract
 */
P2PMetadataEntity = /** @class */ (function (_super) {
    tslib_1.__extends(P2PMetadataEntity, _super);
    function P2PMetadataEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    P2PMetadataEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.sourceAppCode = null;
        this.sourceSuCode = null;
        this.sourceBoId = null;
        this.sourceBeMetadataId = null;
        this.sourceBeId = null;
        //this.sourceBeType = null;
        this.sourceVoMetadataId = null;
        this.sourceVoId = null;
        //this.sourceVoType = null;
        this.targetAppCode = null;
        this.targetSuCode = null;
        this.targetBoId = null;
        this.targetBeMetadataId = null;
        this.targetBeId = null;
        //this.targetBeType = null;
        this.targetVoMetadataId = null;
        this.targetVoId = null;
        //this.targetVoType = null;
        this.targetVoProxyMetadataId = null;
        this.targetVoProxyId = null;
        //this.targetVoProxyType = null;
        this.sourceBeMetadataRootId = null;
        this.sourceBeMetadataParentId = null;
        this.sourceBeMetadataIdPath = null;
        this.sourceBeRootId = null;
        this.sourceBeParentId = null;
        this.sourceBeIdPath = null;
        this.sourceVoMetadataRootId = null;
        this.sourceVoMetadataParentId = null;
        this.sourceVoMetadataIdPath = null;
        this.sourceVoRootId = null;
        this.sourceVoParentId = null;
        this.sourceVoIdPath = null;
        this.targetBeMetadataRootId = null;
        this.targetBeMetadataParentId = null;
        this.targetBeMetadataIdPath = null;
        this.targetBeRootId = null;
        this.targetBeParentId = null;
        this.targetBeIdPath = null;
        this.targetVoMetadataRootId = null;
        this.targetVoMetadataParentId = null;
        this.targetVoMetadataIdPath = null;
        this.targetVoRootId = null;
        this.targetVoParentId = null;
        this.targetVoIdPath = null;
        this.targetVoProxyMetadataRootId = null;
        this.targetVoProxyMetadataParentId = null;
        this.targetVoProxyMetadataIdPath = null;
        this.targetVoProxyRootId = null;
        this.targetVoProxyParentId = null;
        this.targetVoProxyIdPath = null;
    };
    /**
     * @return {?}
     */
    P2PMetadataEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceAppCode", this.sourceAppCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceSuCode", this.sourceSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBoId", this.sourceBoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBeMetadataId", this.sourceBeMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBeId", this.sourceBeId);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBeType", this.sourceBeType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoMetadataId", this.sourceVoMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoType", this.sourceVoType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetAppCode", this.targetAppCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetSuCode", this.targetSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetBoId", this.targetBoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetBeMetadataId", this.targetBeMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetBeId", this.targetBeId);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "targetBeType", this.targetBeType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoMetadataId", this.targetVoMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoId", this.targetVoId);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoType", this.targetVoType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyMetadataId", this.targetVoProxyMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyId", this.targetVoProxyId);
        //BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyType", this.targetVoProxyType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBeMetadataRootId", this.sourceBeMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBeMetadataParentId", this.sourceBeMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBeMetadataIdPath", this.sourceBeMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBeRootId", this.sourceBeRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBeParentId", this.sourceBeParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBeIdPath", this.sourceBeIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoMetadataRootId", this.sourceVoMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoMetadataParentId", this.sourceVoMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoMetadataIdPath", this.sourceVoMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoRootId", this.sourceVoRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoParentId", this.sourceVoParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoIdPath", this.sourceVoIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetBeMetadataRootId", this.targetBeMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetBeMetadataParentId", this.targetBeMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetBeMetadataIdPath", this.targetBeMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetBeRootId", this.targetBeRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetBeParentId", this.targetBeParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetBeIdPath", this.targetBeIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoMetadataRootId", this.targetVoMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoMetadataParentId", this.targetVoMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoMetadataIdPath", this.targetVoMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoRootId", this.targetVoRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoParentId", this.targetVoParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoIdPath", this.targetVoIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyMetadataRootId", this.targetVoProxyMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyMetadataParentId", this.targetVoProxyMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyMetadataIdPath", this.targetVoProxyMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyRootId", this.targetVoProxyRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyParentId", this.targetVoProxyParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyIdPath", this.targetVoProxyIdPath);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    P2PMetadataEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.sourceAppCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceAppCode");
        this.sourceSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceSuCode");
        this.sourceBoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBoId");
        this.sourceBeMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBeMetadataId");
        this.sourceBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBeId");
        //this.sourceBeType= BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBeType");
        this.sourceVoMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoMetadataId");
        this.sourceVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
        //this.sourceVoType= BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoType");
        this.targetAppCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetAppCode");
        this.targetSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetSuCode");
        this.targetBoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBoId");
        this.targetBeMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBeMetadataId");
        this.targetBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBeId");
        //this.targetBeType= BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBeType");
        this.targetVoMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoMetadataId");
        this.targetVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoId");
        //this.targetVoType= BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoType");
        this.targetVoProxyMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyMetadataId");
        this.targetVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyId");
        //this.targetVoProxyType= BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyType");
        this.sourceBeMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBeMetadataRootId");
        this.sourceBeMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBeMetadataParentId");
        this.sourceBeMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBeMetadataIdPath");
        this.sourceBeRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBeRootId");
        this.sourceBeParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBeParentId");
        this.sourceBeIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBeIdPath");
        this.sourceVoMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoMetadataRootId");
        this.sourceVoMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoMetadataParentId");
        this.sourceVoMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoMetadataIdPath");
        this.sourceVoRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoRootId");
        this.sourceVoParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoParentId");
        this.sourceVoIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoIdPath");
        this.targetBeMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBeMetadataRootId");
        this.targetBeMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBeMetadataParentId");
        this.targetBeMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBeMetadataIdPath");
        this.targetBeRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBeRootId");
        this.targetBeParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBeParentId");
        this.targetBeIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBeIdPath");
        this.targetVoMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoMetadataRootId");
        this.targetVoMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoMetadataParentId");
        this.targetVoMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoMetadataIdPath");
        this.targetVoRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoRootId");
        this.targetVoParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoParentId");
        this.targetVoIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoIdPath");
        this.targetVoProxyMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyMetadataRootId");
        this.targetVoProxyMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyMetadataParentId");
        this.targetVoProxyMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyMetadataIdPath");
        this.targetVoProxyRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyRootId");
        this.targetVoProxyParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyParentId");
        this.targetVoProxyIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyIdPath");
    };
    return P2PMetadataEntity;
}(MetadataEntity));
/**
 * 点到点元数据实体类。
 * @abstract
 */
export { P2PMetadataEntity };
if (false) {
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceAppCode;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceSuCode;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceBoId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceBeMetadataId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceBeId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceVoMetadataId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceVoId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetAppCode;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetSuCode;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetBoId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetBeMetadataId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetBeId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoMetadataId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoProxyMetadataId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoProxyId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceBeMetadataRootId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceBeMetadataParentId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceBeMetadataIdPath;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceBeRootId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceBeParentId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceBeIdPath;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceVoMetadataRootId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceVoMetadataParentId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceVoMetadataIdPath;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceVoRootId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceVoParentId;
    /** @type {?} */
    P2PMetadataEntity.prototype.sourceVoIdPath;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetBeMetadataRootId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetBeMetadataParentId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetBeMetadataIdPath;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetBeRootId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetBeParentId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetBeIdPath;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoMetadataRootId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoMetadataParentId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoMetadataIdPath;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoRootId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoParentId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoIdPath;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoProxyMetadataRootId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoProxyMetadataParentId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoProxyMetadataIdPath;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoProxyRootId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoProxyParentId;
    /** @type {?} */
    P2PMetadataEntity.prototype.targetVoProxyIdPath;
}
