// @edp-aif/common-api/esm5/lib/entity/impl/Participant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from "../../BpmnModelJsonHelper";
import { JsonConvertableImpl } from "../../JsonConvertableImpl";
import { ParticipantType } from "./ParticipantType";
import { PositionParticipant } from "./PositionParticipant";
import { UserParticipant } from "./UserParticipant";
var Participant = /** @class */ (function (_super) {
    tslib_1.__extends(Participant, _super);
    function Participant() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.participantType = ParticipantType.UserParticipant;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    Participant.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.participantType = ParticipantType.UserParticipant;
        this.userParticipant = null;
        this.positionParticipant = null;
    };
    /**
     * @return {?}
     */
    Participant.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "participantType", this.participantType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "userParticipant", this.userParticipant);
        BpmnModelJsonHelper.SetAttributeValue(obj, "positionParticipant", this.positionParticipant);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    Participant.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.participantType = ParticipantType[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "participantType")];
        this.userParticipant = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "userParticipant", UserParticipant);
        this.positionParticipant = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "positionParticipant", PositionParticipant);
    };
    return Participant;
}(JsonConvertableImpl));
export { Participant };
if (false) {
    /** @type {?} */
    Participant.prototype.participantType;
    /** @type {?} */
    Participant.prototype.userParticipant;
    /** @type {?} */
    Participant.prototype.positionParticipant;
}
