// @edp-aif/common-api/esm5/lib/entity/impl/PositionParticipant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from "../../BpmnModelJsonHelper";
import { JsonConvertableImpl } from "../../JsonConvertableImpl";
var PositionParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(PositionParticipant, _super);
    function PositionParticipant() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.orgType = "Sys";
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    PositionParticipant.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.positionId = null;
        this.positionName = null;
        this.positionType = null;
        this.orgId = null;
        this.orgName = null;
        this.orgType = null;
        this.orgIdExpression = null;
    };
    /**
     * @return {?}
     */
    PositionParticipant.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "positionId", this.positionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "positionName", this.positionName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "positionType", this.positionType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgId", this.orgId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgName", this.orgName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgType", this.orgType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgIdExpression", this.orgIdExpression);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    PositionParticipant.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.positionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "positionId");
        this.positionName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "positionName");
        this.positionType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "positionType");
        this.orgId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgId");
        this.orgName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgName");
        this.orgType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgType");
        this.orgIdExpression = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgIdExpression");
    };
    return PositionParticipant;
}(JsonConvertableImpl));
export { PositionParticipant };
if (false) {
    /** @type {?} */
    PositionParticipant.prototype.positionId;
    /** @type {?} */
    PositionParticipant.prototype.positionName;
    /** @type {?} */
    PositionParticipant.prototype.positionType;
    /** @type {?} */
    PositionParticipant.prototype.orgId;
    /** @type {?} */
    PositionParticipant.prototype.orgName;
    /** @type {?} */
    PositionParticipant.prototype.orgType;
    /** @type {?} */
    PositionParticipant.prototype.orgIdExpression;
}
