// @edp-aif/common-api/esm5/lib/entity/impl/TaskHandlingMenuParameter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from "../../BpmnModelJsonHelper";
import { JsonConvertableImpl } from "../../JsonConvertableImpl";
import { TaskHandlingMenuParameterMode } from "./TaskHandlingMenuParameterMode";
var TaskHandlingMenuParameter = /** @class */ (function (_super) {
    tslib_1.__extends(TaskHandlingMenuParameter, _super);
    function TaskHandlingMenuParameter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.parameterMode = TaskHandlingMenuParameterMode.CONSTANT;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    TaskHandlingMenuParameter.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.parameterMode = TaskHandlingMenuParameterMode.CONSTANT;
        this.code = null;
        this.constantValue = null;
        this.fieldContent = null;
        this.expressionContent = null;
    };
    /**
     * @return {?}
     */
    TaskHandlingMenuParameter.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "parameterMode", this.parameterMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "constantValue", this.constantValue);
        BpmnModelJsonHelper.SetAttributeValue(obj, "fieldContent", this.fieldContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "expressionContent", this.expressionContent);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    TaskHandlingMenuParameter.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.parameterMode = TaskHandlingMenuParameterMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "parameterMode")];
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.constantValue = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "constantValue");
        this.fieldContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "fieldContent");
        this.expressionContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "expressionContent");
    };
    return TaskHandlingMenuParameter;
}(JsonConvertableImpl));
export { TaskHandlingMenuParameter };
if (false) {
    /** @type {?} */
    TaskHandlingMenuParameter.prototype.parameterMode;
    /** @type {?} */
    TaskHandlingMenuParameter.prototype.code;
    /** @type {?} */
    TaskHandlingMenuParameter.prototype.constantValue;
    /** @type {?} */
    TaskHandlingMenuParameter.prototype.fieldContent;
    /** @type {?} */
    TaskHandlingMenuParameter.prototype.expressionContent;
}
