// @edp-aif/common-api/esm5/lib/entity/impl/TaskHandlingMenuArgs.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from "../../BpmnModelJsonHelper";
import { JsonConvertableImpl } from "../../JsonConvertableImpl";
import { TaskHandlingMenuParameter } from "./TaskHandlingMenuParameter";
var TaskHandlingMenuArgs = /** @class */ (function (_super) {
    tslib_1.__extends(TaskHandlingMenuArgs, _super);
    function TaskHandlingMenuArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.parameters = [];
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    TaskHandlingMenuArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.menuId = null;
        this.menuName = null;
        this.parameters = [];
    };
    /**
     * @return {?}
     */
    TaskHandlingMenuArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "menuId", this.menuId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "menuName", this.menuName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "parameters", this.parameters);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    TaskHandlingMenuArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.menuId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "menuId");
        this.menuName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "menuName");
        this.parameters = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "parameters", TaskHandlingMenuParameter);
    };
    return TaskHandlingMenuArgs;
}(JsonConvertableImpl));
export { TaskHandlingMenuArgs };
if (false) {
    /** @type {?} */
    TaskHandlingMenuArgs.prototype.menuId;
    /** @type {?} */
    TaskHandlingMenuArgs.prototype.menuName;
    /** @type {?} */
    TaskHandlingMenuArgs.prototype.parameters;
}
