// @edp-aif/common-api/esm5/lib/entity/impl/TaskCenterContent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from "../../BpmnModelJsonHelper";
import { JsonConvertableImpl } from "../../JsonConvertableImpl";
import { Participant } from "./Participant";
import { TaskHandlingMenuArgs } from "./TaskHandlingMenuArgs";
var TaskCenterContent = /** @class */ (function (_super) {
    tslib_1.__extends(TaskCenterContent, _super);
    function TaskCenterContent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.participants = [];
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    TaskCenterContent.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.participants = [];
        this.taskHandlingMenuArgs = null;
    };
    /**
     * @return {?}
     */
    TaskCenterContent.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "participants", this.participants);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskHandlingMenuArgs", this.taskHandlingMenuArgs);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    TaskCenterContent.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.participants = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "participants", Participant);
        this.taskHandlingMenuArgs = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "taskHandlingMenuArgs", TaskHandlingMenuArgs);
    };
    return TaskCenterContent;
}(JsonConvertableImpl));
export { TaskCenterContent };
if (false) {
    /** @type {?} */
    TaskCenterContent.prototype.participants;
    /** @type {?} */
    TaskCenterContent.prototype.taskHandlingMenuArgs;
}
