// @edp-aif/common-api/esm5/lib/entity/impl/UserParticipant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from "../../BpmnModelJsonHelper";
import { JsonConvertableImpl } from "../../JsonConvertableImpl";
var UserParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(UserParticipant, _super);
    function UserParticipant() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.orgType = "Sys";
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    UserParticipant.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.userId = null;
        this.userName = null;
        this.orgId = null;
        this.orgName = null;
        this.orgType = null;
    };
    /**
     * @return {?}
     */
    UserParticipant.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "userId", this.userId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "userName", this.userName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgId", this.orgId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgName", this.orgName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "orgType", this.orgType);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    UserParticipant.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.userId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "userId");
        this.userName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "userName");
        this.orgId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgId");
        this.orgName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgName");
        this.orgType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "orgType");
    };
    return UserParticipant;
}(JsonConvertableImpl));
export { UserParticipant };
if (false) {
    /** @type {?} */
    UserParticipant.prototype.userId;
    /** @type {?} */
    UserParticipant.prototype.userName;
    /** @type {?} */
    UserParticipant.prototype.orgId;
    /** @type {?} */
    UserParticipant.prototype.orgName;
    /** @type {?} */
    UserParticipant.prototype.orgType;
}
