// @farris/mobile-cli/lib/config/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.baseOutputOption = exports.baseInputOption = void 0;
const plugin_node_resolve_1 = __importDefault(require("@rollup/plugin-node-resolve"));
const plugin_commonjs_1 = __importDefault(require("@rollup/plugin-commonjs"));
const rollup_plugin_terser_1 = require("rollup-plugin-terser");
const plugin_replace_1 = __importDefault(require("@rollup/plugin-replace"));
const utils_1 = require("../utils");
const plugins = [
    plugin_node_resolve_1.default(),
    plugin_commonjs_1.default(),
    plugin_replace_1.default({ 'process.env.NODE_ENV': JSON.stringify(utils_1.ENV) }),
];
if (utils_1.ENV === 'production') {
    plugins.push(rollup_plugin_terser_1.terser({ format: { comments: false } }));
}
exports.baseInputOption = {
    plugins,
};
exports.baseOutputOption = {
    file: 'dist/index.js',
    format: 'system',
};
exports.default = {
    input: exports.baseInputOption,
    output: exports.baseOutputOption,
};
