// @farris/mobile-cli/lib/config/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkLib = exports.libs = exports.getOption = exports.getOutputOption = exports.getInputOption = void 0;
const utils_1 = require("../utils");
const base_1 = require("./base");
const project_1 = require("./project");
const libs_1 = require("./libs");
const inputOptions = Object.assign({ base: base_1.baseInputOption, project: project_1.projectInputOption }, libs_1.libInputOptions);
const outputOptions = Object.assign({ base: base_1.baseOutputOption, project: project_1.projectOutputOption }, libs_1.libOutputOptions);
exports.getInputOption = (type, option) => {
    let resultOptions = {};
    if (type) {
        resultOptions = inputOptions[type] ? utils_1.merge(inputOptions.base, inputOptions[type]) : inputOptions.base;
    }
    return utils_1.merge(resultOptions, option ? option : {});
};
exports.getOutputOption = (type, option) => {
    let resultOptions = {};
    if (type) {
        resultOptions = outputOptions[type] ? utils_1.merge(outputOptions.base, outputOptions[type]) : outputOptions.base;
    }
    return utils_1.merge(resultOptions, option ? option : {});
};
exports.getOption = (option) => {
    if (option.type) {
        return {
            input: exports.getInputOption(option.type, option.input),
            output: exports.getOutputOption(option.type, option.out)
        };
    }
    return {
        input: utils_1.merge(base_1.baseInputOption, option),
        output: utils_1.merge(base_1.baseOutputOption, option)
    };
};
exports.libs = [
    'vue',
    'vue-router',
    'dayjs',
    'axios',
    'vant',
    'quill',
    'vue-compiler'
];
exports.checkLib = (lib) => {
    return outputOptions[lib] ? true : false;
};
