// @farris/bef/esm5/lib/deprecated/deprecated_bef_repository.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/deprecated/deprecated_bef_repository.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * @Author: Sagi, Lucas, Witt
 * @Date: 2018-10-11 17:34:26
 * @Last Modified by: aalizzwell
 * @Last Modified time: 2019-06-29 11:39:02
 */
// tslint:disable: max-line-length
import { Injectable } from '@angular/core';
import { of } from 'rxjs';
import { Repository } from '@farris/devkit';
/**
 * 适配老的API，保持兼容
 * @abstract
 * @template T
 */
var DeprecatedBefRepository = /** @class */ (function (_super) {
    tslib_1.__extends(DeprecatedBefRepository, _super);
    // #endregion
    /**
    * 构造函数
    */
    function DeprecatedBefRepository() {
        return _super.call(this) || this;
    }
    Object.defineProperty(DeprecatedBefRepository.prototype, "restService", {
        // #region 老的属性包装成get方法进行兼容
        /**
         * 取数服务
         */
        get: 
        // #region 老的属性包装成get方法进行兼容
        /**
         * 取数服务
         * @return {?}
         */
        function () {
            return this.dataService.restService;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DeprecatedBefRepository.prototype, "changeBuilder", {
        /**
         * 实体变更（前端）=> BE变更（后端）
         */
        get: /**
         * 实体变更（前端）=> BE变更（后端）
         * @return {?}
         */
        function () {
            return this.entityManager.changeBuilder;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DeprecatedBefRepository.prototype, "changeHandler", {
        /**
         * BE变更（后端）=> 实体变更（前端）
         */
        get: /**
         * BE变更（后端）=> 实体变更（前端）
         * @return {?}
         */
        function () {
            return this.entityManager.changeHandler;
        },
        enumerable: true,
        configurable: true
    });
    // #region 数据访问
    /**
    * 获取实体集合
    */
    // #region 数据访问
    /**
     * 获取实体集合
     * @return {?}
     */
    DeprecatedBefRepository.prototype.getList = 
    // #region 数据访问
    /**
     * 获取实体集合
     * @return {?}
     */
    function () {
        return this.dataService.getList(null, null, null, null);
    };
    /**
     * 获取单个实体
     * @param id 主实体id
     */
    /**
     * 获取单个实体
     * @param {?} id 主实体id
     * @return {?}
     */
    DeprecatedBefRepository.prototype.getById = /**
     * 获取单个实体
     * @param {?} id 主实体id
     * @return {?}
     */
    function (id) {
        return this.dataService.getById(id);
    };
    /**
     * 根据id更新实体
     * @param id 实体id
     */
    /**
     * 根据id更新实体
     * @param {?} id 实体id
     * @return {?}
     */
    DeprecatedBefRepository.prototype.updateById = /**
     * 根据id更新实体
     * @param {?} id 实体id
     * @return {?}
     */
    function (id) {
        return this.dataService.updateById(id);
    };
    /**
     * 创建新实体，并加载
     */
    /**
     * 创建新实体，并加载
     * @param {?=} defaultValue
     * @return {?}
     */
    DeprecatedBefRepository.prototype.create = /**
     * 创建新实体，并加载
     * @param {?=} defaultValue
     * @return {?}
     */
    function (defaultValue) {
        return this.dataService.create(defaultValue);
    };
    /**
     * 追加实体
     */
    /**
     * 追加实体
     * @param {?=} defaultValue
     * @return {?}
     */
    DeprecatedBefRepository.prototype.append = /**
     * 追加实体
     * @param {?=} defaultValue
     * @return {?}
     */
    function (defaultValue) {
        return this.dataService.append(defaultValue);
    };
    /**
     * 创建子实体，并追加
     * @param fpath 父路径，格式形如：/1/edus（从表）或/1/edus/2/grades（从从表）
     */
    /**
     * 创建子实体，并追加
     * @param {?} fpath 父路径，格式形如：/1/edus（从表）或/1/edus/2/grades（从从表）
     * @return {?}
     */
    DeprecatedBefRepository.prototype.appendByPath = /**
     * 创建子实体，并追加
     * @param {?} fpath 父路径，格式形如：/1/edus（从表）或/1/edus/2/grades（从从表）
     * @return {?}
     */
    function (fpath) {
        return this.dataService.appendByPath(fpath);
    };
    /**
     * 根据id删除实体
     * @param id 内码
     */
    /**
     * 根据id删除实体
     * @param {?} id 内码
     * @param {?=} ifSave
     * @return {?}
     */
    DeprecatedBefRepository.prototype.removeById = /**
     * 根据id删除实体
     * @param {?} id 内码
     * @param {?=} ifSave
     * @return {?}
     */
    function (id, ifSave) {
        return this.dataService.removeById(id, ifSave);
    };
    /**
     * 批量删除
     * @param id 内码
     * @param ifSave 是否保存
     */
    /**
     * 批量删除
     * @param {?} ids
     * @param {?=} ifSave 是否保存
     * @return {?}
     */
    DeprecatedBefRepository.prototype.removeByIds = /**
     * 批量删除
     * @param {?} ids
     * @param {?=} ifSave 是否保存
     * @return {?}
     */
    function (ids, ifSave) {
        return this.dataService.removeByIds(ids, ifSave);
    };
    /**
     * 删除子级
     * @param path 路径
     * @param id   内码
     */
    /**
     * 删除子级
     * @param {?} fpath
     * @param {?} id   内码
     * @return {?}
     */
    DeprecatedBefRepository.prototype.removeByPath = /**
     * 删除子级
     * @param {?} fpath
     * @param {?} id   内码
     * @return {?}
     */
    function (fpath, id) {
        return this.dataService.removeByPath(fpath, id);
    };
    /**
     * 将id对应的实体的变更提交的服务器端
     */
    /**
     * 将id对应的实体的变更提交的服务器端
     * @param {?} id
     * @return {?}
     */
    DeprecatedBefRepository.prototype.updateChangesById = /**
     * 将id对应的实体的变更提交的服务器端
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.dataService.updateChangesById(id);
    };
    /**
     * 提交path对应实体的变更
     */
    /**
     * 提交path对应实体的变更
     * @param {?} path
     * @param {?} id
     * @return {?}
     */
    DeprecatedBefRepository.prototype.updateChangesByPath = /**
     * 提交path对应实体的变更
     * @param {?} path
     * @param {?} id
     * @return {?}
     */
    function (path, id) {
        throw new Error('Not Implement');
    };
    /**
     * 提交所有变更
     */
    /**
     * 提交所有变更
     * @return {?}
     */
    DeprecatedBefRepository.prototype.updateAllChanges = /**
     * 提交所有变更
     * @return {?}
     */
    function () {
        return this.dataService.updateAllChanges();
    };
    /**
     * 应用变更
     */
    /**
     * 应用变更
     * @return {?}
     */
    DeprecatedBefRepository.prototype.applyChanges = /**
     * 应用变更
     * @return {?}
     */
    function () {
        return this.dataService.applyChanges();
    };
    /**
     * 提交数据变更和变量变更，通过retrieve模拟
     * @internal
     */
    /**
     * 提交数据变更和变量变更，通过retrieve模拟
     * \@internal
     * @return {?}
     */
    DeprecatedBefRepository.prototype.updateDataAndVariableChanges = /**
     * 提交数据变更和变量变更，通过retrieve模拟
     * \@internal
     * @return {?}
     */
    function () {
        /** @type {?} */
        var filterConditions = [];
        /** @type {?} */
        var sortConditions = [];
        /** @type {?} */
        var entityFilter = {
            FilterConditions: filterConditions,
            SortConditions: sortConditions,
            IsUsePagination: true,
            Pagination: {
                PageIndex: 1,
                PageSize: 1,
                PageCount: 0,
                TotalCount: 0
            }
        };
        /** @type {?} */
        var entityFilterJson = JSON.stringify(entityFilter);
        /** @type {?} */
        var requestInfo = this.restService.buildRequestInfo();
        // 没有变更时不再提交变更
        if (this.isEmptyRequestInfo(requestInfo)) {
            return of(null);
        }
        /** @type {?} */
        var url = this.restService.baseUri + "/extension/query?entityFilter=" + entityFilterJson;
        /** @type {?} */
        var body = requestInfo;
        /** @type {?} */
        var options = {
            body: body
        };
        return this.restService.invoke(url, 'PUT', null, options, false, true, true);
    };
    /**
     * 应用id对应数据的变更
     */
    /**
     * 应用id对应数据的变更
     * @param {?} id
     * @return {?}
     */
    DeprecatedBefRepository.prototype.applyChangesById = /**
     * 应用id对应数据的变更
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.dataService.applyChangesById(id);
    };
    /**
     * 取消变更
     */
    /**
     * 取消变更
     * @param {?=} options
     * @return {?}
     */
    DeprecatedBefRepository.prototype.cancelChanges = /**
     * 取消变更
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        return this.dataService.cancelChanges(options);
    };
    // #endregion
    // #region 数据变更集处理
    /**
     * 清空所有本地实体的变更集
     */
    // #endregion
    // #region 数据变更集处理
    /**
     * 清空所有本地实体的变更集
     * @return {?}
     */
    DeprecatedBefRepository.prototype.clearAllEntityChanges = 
    // #endregion
    // #region 数据变更集处理
    /**
     * 清空所有本地实体的变更集
     * @return {?}
     */
    function () {
        return this.entityManager.clearAllEntityChanges();
    };
    /**
     * 获取数据变更
     */
    /**
     * 获取数据变更
     * @return {?}
     */
    DeprecatedBefRepository.prototype.getDataChangeDetails = /**
     * 获取数据变更
     * @return {?}
     */
    function () {
        return this.entityManager.buildAllEntityChangeDetails();
    };
    /**
     * 处理数据变更
     */
    /**
     * 处理数据变更
     * @param {?} changeDetails
     * @return {?}
     */
    DeprecatedBefRepository.prototype.handleDataChangeDetails = /**
     * 处理数据变更
     * @param {?} changeDetails
     * @return {?}
     */
    function (changeDetails) {
        return this.entityManager.handleDataChangeDetails(changeDetails);
    };
    // #endregion
    // #region 变量变更集处理
    /**
     * 获取变量变更
     */
    // #endregion
    // #region 变量变更集处理
    /**
     * 获取变量变更
     * @return {?}
     */
    DeprecatedBefRepository.prototype.getVariableChangeDetail = 
    // #endregion
    // #region 变量变更集处理
    /**
     * 获取变量变更
     * @return {?}
     */
    function () {
        return this.variableManager.buildChangeDetail();
    };
    /**
     * 处理变量变更
     */
    /**
     * 处理变量变更
     * @param {?} changeDetail
     * @return {?}
     */
    DeprecatedBefRepository.prototype.handleVariableChangeDetail = /**
     * 处理变量变更
     * @param {?} changeDetail
     * @return {?}
     */
    function (changeDetail) {
        this.variableManager.handleChangeDetail(changeDetail);
    };
    // #endregion
    /**
     * 变更集是否为空
     * @param requestInfo requestInfo
     */
    // #endregion
    /**
     * 变更集是否为空
     * @private
     * @param {?} requestInfo requestInfo
     * @return {?}
     */
    DeprecatedBefRepository.prototype.isEmptyRequestInfo = 
    // #endregion
    /**
     * 变更集是否为空
     * @private
     * @param {?} requestInfo requestInfo
     * @return {?}
     */
    function (requestInfo) {
        if (!requestInfo) {
            return true;
        }
        if ((!requestInfo.dataChange || requestInfo.dataChange.length < 1) && (!requestInfo.variableChange || Object.keys(requestInfo.variableChange).length < 1)) {
            return true;
        }
        return false;
    };
    DeprecatedBefRepository.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DeprecatedBefRepository.ctorParameters = function () { return []; };
    return DeprecatedBefRepository;
}(Repository));
if (false) {
    /**
     * 实体管理器
     * \@internal
     * @type {?}
     */
    DeprecatedBefRepository.prototype.entityManager;
    /**
     * 变量管理器
     * \@internal
     * @type {?}
     */
    DeprecatedBefRepository.prototype.variableManager;
    /**
     * 取数服务
     * \@internal
     * @type {?}
     * @protected
     */
    DeprecatedBefRepository.prototype.dataService;
}
export { DeprecatedBefRepository };
