// @gsp-cmp/ubcc-param-data-ui/esm5/lib/ubcc-param-dimension-ui-root/ubcc-param-dimension-ui-root.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ubcc-param-dimension-ui-root/ubcc-param-dimension-ui-root.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { UbccParamTemplate } from '@gsp-cmp/ubcc-param-vo';
var UbccParamDimensionUiRootComponent = /** @class */ (function () {
    function UbccParamDimensionUiRootComponent() {
        this.dimensionItemDatas = [];
        this.paramDataItems = [];
        this.canEdit = true;
    }
    /**
     * @return {?}
     */
    UbccParamDimensionUiRootComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.template = this.template || new UbccParamTemplate();
    };
    UbccParamDimensionUiRootComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-ubcc-param-dimension-ui-root',
                    template: "\n<lib-ubcc-param-dimension-ui [(dimensionItemDatas)]=\"dimensionItemDatas\"></lib-ubcc-param-dimension-ui>\n<lib-ubcc-param-data-ui [canEdit]=\"canEdit\" [(template)]=\"template\" [(paramDataItems)]=\"paramDataItems\"></lib-ubcc-param-data-ui>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    UbccParamDimensionUiRootComponent.ctorParameters = function () { return []; };
    UbccParamDimensionUiRootComponent.propDecorators = {
        template: [{ type: Input }],
        dimensionItemDatas: [{ type: Input }],
        paramDataItems: [{ type: Input }],
        canEdit: [{ type: Input }]
    };
    return UbccParamDimensionUiRootComponent;
}());
export { UbccParamDimensionUiRootComponent };
if (false) {
    /** @type {?} */
    UbccParamDimensionUiRootComponent.prototype.template;
    /** @type {?} */
    UbccParamDimensionUiRootComponent.prototype.dimensionItemDatas;
    /** @type {?} */
    UbccParamDimensionUiRootComponent.prototype.paramDataItems;
    /** @type {?} */
    UbccParamDimensionUiRootComponent.prototype.canEdit;
}
