// @edp-bif/runtime-api/esm5/lib/bif-manual-creation/args/GetEntityDatasArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bif-manual-creation/args/GetEntityDatasArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
var GetEntityDatasArgs = /** @class */ (function (_super) {
    tslib_1.__extends(GetEntityDatasArgs, _super);
    function GetEntityDatasArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    GetEntityDatasArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.voId = null;
        this.dataId = null;
        this.suCode = null;
    };
    /**
     * @return {?}
     */
    GetEntityDatasArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaId", this.voId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dataId", this.dataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    GetEntityDatasArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaId");
        this.dataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dataId");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
    };
    return GetEntityDatasArgs;
}(JsonConvertableImpl));
export { GetEntityDatasArgs };
if (false) {
    /** @type {?} */
    GetEntityDatasArgs.prototype.voId;
    /** @type {?} */
    GetEntityDatasArgs.prototype.dataId;
    /** @type {?} */
    GetEntityDatasArgs.prototype.suCode;
}
