// @edp-bif/runtime-api/esm5/lib/bif-manual-creation/args/PushTaskArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bif-manual-creation/args/PushTaskArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { RpcCreationRuleExecutorResult } from "@edp-aif/runtime-api";
var PushTaskArgs = /** @class */ (function (_super) {
    tslib_1.__extends(PushTaskArgs, _super);
    function PushTaskArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    PushTaskArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.suCode = null;
        this.creationRuleExecutorResult = null;
    };
    /**
     * @return {?}
     */
    PushTaskArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleExecutorResult", this.creationRuleExecutorResult);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    PushTaskArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.creationRuleExecutorResult = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "creationRuleExecutorResult", RpcCreationRuleExecutorResult);
    };
    return PushTaskArgs;
}(JsonConvertableImpl));
export { PushTaskArgs };
if (false) {
    /** @type {?} */
    PushTaskArgs.prototype.suCode;
    /** @type {?} */
    PushTaskArgs.prototype.creationRuleExecutorResult;
}
