// @farris/ui-personnel-selector/esm5/lib/selection.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/selection.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from "@angular/core";
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
var SelectionService = /** @class */ (function () {
    function SelectionService() {
        this.showSubject = new Subject();
    }
    /**
     * @return {?}
     */
    SelectionService.prototype.show = /**
     * @return {?}
     */
    function () {
        this.showSubject.next(true);
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    SelectionService.prototype.set = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        localStorage.setItem(key, JSON.stringify(value)); // 只能是字符串
    };
    /**
     * @param {?} key
     * @return {?}
     */
    SelectionService.prototype.get = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return JSON.parse(localStorage.getItem(key));
    };
    /**
     * @param {?} key
     * @return {?}
     */
    SelectionService.prototype.remove = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        localStorage.removeItem(key);
    };
    SelectionService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    SelectionService.ctorParameters = function () { return []; };
    /** @nocollapse */ SelectionService.ngInjectableDef = i0.defineInjectable({ factory: function SelectionService_Factory() { return new SelectionService(); }, token: SelectionService, providedIn: "root" });
    return SelectionService;
}());
export { SelectionService };
if (false) {
    /** @type {?} */
    SelectionService.prototype.showSubject;
}
