// @farris/devkit/esm2015/lib/variable/frame_id_variable_parser.js
/**
 * session变量解析
 * @author Witt <jiwt@inspur.com>
 */
import { Injectable } from '@angular/core';
import { ParseUtil } from './parse_util';
const VARIABLE_MATCHER = /#{\S+?}/g;
/**
 * 数据变量解析
 */
class FrameIdVariableParser {
    /**
     * 解析变量
     * @param expression 形如：/#{root-component}/stateName
     * @param context 上下文
     */
    parse(expression, context) {
        this.context = context;
        const variables = this.extractVariables(expression);
        if (!variables || variables.length < 1) {
            return expression;
        }
        variables.forEach((exp) => {
            const frameId = this.getVariableValue(exp);
            expression = expression.replace(VARIABLE_MATCHER, frameId);
        });
        return expression;
    }
    /**
     * 计算表达式的值
     * @param expression like #{root-component}
     */
    getVariableValue(expression) {
        const relativeFrameId = expression.substring(2, expression.length - 1);
        return this.getFullFrameId(relativeFrameId);
    }
    extractVariables(expression) {
        if (!expression) {
            return [];
        }
        return expression.match(VARIABLE_MATCHER);
    }
    /**
     * 获取补全的frameId
     * @param frameId frameId
     */
    getFullFrameId(frameId) {
        const frameContext = ParseUtil.getFrameContext(this.context);
        const namespace = frameContext.namespace || '';
        const prefix = namespace ? `${namespace}_` : '';
        return `${prefix}${frameId}`;
    }
}
FrameIdVariableParser.decorators = [
    { type: Injectable }
];
export { FrameIdVariableParser };
