// @farris/ide-devkit/esm5/lib/ideframework/gsp/gsp-event-bus.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { EventBus } from '../eventbus';
var GspEventBus = /** @class */ (function (_super) {
    tslib_1.__extends(GspEventBus, _super);
    function GspEventBus() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.commands = [];
        return _this;
    }
    /**
     * @param {?} target
     * @param {?} tokenValue
     * @param {?} eventName
     * @param {?} caller
     * @param {?} handler
     * @return {?}
     */
    GspEventBus.prototype.on = /**
     * @param {?} target
     * @param {?} tokenValue
     * @param {?} eventName
     * @param {?} caller
     * @param {?} handler
     * @return {?}
     */
    function (target, tokenValue, eventName, caller, handler) {
        // 如果有对应激活命令，监听事件时立即执行回调函数
        /** @type {?} */
        var command = this.commands.find(function (item) { return item.name === eventName; });
        if (command) {
            handler.call(caller, command.eventeParams);
        }
        return _super.prototype.on.call(this, target, tokenValue, eventName, caller, handler);
    };
    /**
     * @param {?} emitterType
     * @param {?} tokenValue
     * @param {?} eventName
     * @param {?} eventArgs
     * @return {?}
     */
    GspEventBus.prototype.post = /**
     * @param {?} emitterType
     * @param {?} tokenValue
     * @param {?} eventName
     * @param {?} eventArgs
     * @return {?}
     */
    function (emitterType, tokenValue, eventName, eventArgs) {
        this.dispatchCommand(eventName, eventArgs);
        _super.prototype.post.call(this, emitterType, tokenValue, eventName, eventArgs);
    };
    /**
     * @param {?} commandName
     * @param {?} plugin
     * @param {?} activateFn
     * @return {?}
     */
    GspEventBus.prototype.registryActivateCommand = /**
     * @param {?} commandName
     * @param {?} plugin
     * @param {?} activateFn
     * @return {?}
     */
    function (commandName, plugin, activateFn) {
        this.commands.push({
            name: commandName,
            plugin: plugin,
            activateFn: activateFn
        });
    };
    /**
     * @private
     * @param {?} commandName
     * @param {?} args
     * @return {?}
     */
    GspEventBus.prototype.dispatchCommand = /**
     * @private
     * @param {?} commandName
     * @param {?} args
     * @return {?}
     */
    function (commandName, args) {
        /** @type {?} */
        var command = this.commands.find(function (item) { return item.name === commandName; });
        if (command) {
            command.activateFn(command.plugin);
            command.eventParams = args;
        }
    };
    return GspEventBus;
}(EventBus));
export { GspEventBus };
if (false) {
    /**
     * @type {?}
     * @private
     */
    GspEventBus.prototype.commands;
}
