// @farris/ide-devkit/esm5/lib/ideframework/gsp/gsp.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CacheService } from './../../services/cache/services/cache.service';
import { CommandRegistry } from './command-registry';
import { Util } from '../util/util';
import { ViewRegistry } from './view-registry';
import { Workspace } from './workspace';
import { GspEventBus } from './gsp-event-bus';
import { Ide } from './ide/ide';
var GSP = /** @class */ (function () {
    function GSP(parent) {
        this.pConfig = {
            get: /**
             * @param {?} key
             * @return {?}
             */
            function (key) {
                return this[key];
            }
        };
        /** @type {?} */
        var ide;
        if (parent) {
            this.pViews = parent.views;
            this.pCommands = parent.commands;
            this.pEventBus = parent.eventBus;
            this.pUtil = parent.util;
            this.pWorkspace = parent.workspace;
            this.pCache = parent.cache;
            ide = parent.ide;
            /** @type {?} */
            var ctor = parent.context.constructor;
            this.pContext = new ctor();
        }
        else {
            this.pViews = new ViewRegistry(this);
            this.pCommands = new CommandRegistry();
            this.pEventBus = new GspEventBus();
            this.pUtil = new Util();
            this.pWorkspace = new Workspace({
                config: this.pConfig,
                viewRegistry: this.pViews
            });
            this.pCache = new CacheService(null);
        }
        this.pIde = new Ide(ide);
    }
    Object.defineProperty(GSP.prototype, "eventBus", {
        get: /**
         * @return {?}
         */
        function () { return this.pEventBus; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GSP.prototype, "util", {
        get: /**
         * @return {?}
         */
        function () { return this.pUtil; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GSP.prototype, "workspace", {
        get: /**
         * @return {?}
         */
        function () { return this.pWorkspace; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GSP.prototype, "views", {
        get: /**
         * @return {?}
         */
        function () { return this.pViews; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GSP.prototype, "commands", {
        get: /**
         * @return {?}
         */
        function () { return this.pCommands; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GSP.prototype, "ide", {
        get: /**
         * @return {?}
         */
        function () { return this.pIde; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GSP.prototype, "cache", {
        get: /**
         * @return {?}
         */
        function () { return this.pCache; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GSP.prototype, "context", {
        get: /**
         * @return {?}
         */
        function () { return this.pContext; },
        enumerable: true,
        configurable: true
    });
    /**
     * @deprecated 由子模块独自创建gsp实例，传入父实例作为参数。
     * 每个iframe持有一个独有的gsp实例，其中eventbus共享，ide不共享
     */
    /**
     * @deprecated 由子模块独自创建gsp实例，传入父实例作为参数。
     * 每个iframe持有一个独有的gsp实例，其中eventbus共享，ide不共享
     * @return {?}
     */
    GSP.prototype.createChild = /**
     * @deprecated 由子模块独自创建gsp实例，传入父实例作为参数。
     * 每个iframe持有一个独有的gsp实例，其中eventbus共享，ide不共享
     * @return {?}
     */
    function () {
        /** @type {?} */
        var child = new GSP(this);
        // child.pEventBus = this.pEventBus;
        // child.pViews = this.pViews;
        // child.pWorkspace = this.pWorkspace;
        return child;
    };
    return GSP;
}());
export { GSP };
if (false) {
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pEventBus;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pUtil;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pWorkspace;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pConfig;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pViews;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pCommands;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pIde;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pCache;
    /**
     * @type {?}
     * @private
     */
    GSP.prototype.pContext;
}
