// @farris/ide-devkit/esm5/lib/ideframework/gsp/pane-axis.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Model } from './model';
var PaneAxis = /** @class */ (function (_super) {
    tslib_1.__extends(PaneAxis, _super);
    function PaneAxis(params, viewRegistry) {
        return _super.call(this, {}) || this;
    }
    /**
     * @return {?}
     */
    PaneAxis.prototype.deserialize = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    PaneAxis.prototype.getElement = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    PaneAxis.prototype.addChild = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    PaneAxis.prototype.removeChidl = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    PaneAxis.prototype.insertChildBefore = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    PaneAxis.prototype.subscribeToChild = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    PaneAxis.prototype.unsubscribeFromChild = /**
     * @return {?}
     */
    function () { };
    return PaneAxis;
}(Model));
export { PaneAxis };
if (false) {
    /** @type {?} */
    PaneAxis.prototype.flexScale;
    /** @type {?} */
    PaneAxis.prototype.parent;
    /** @type {?} */
    PaneAxis.prototype.container;
    /** @type {?} */
    PaneAxis.prototype.orientation;
    /** @type {?} */
    PaneAxis.prototype.children;
    /** @type {?} */
    PaneAxis.prototype.panes;
    /** @type {?} */
    PaneAxis.prototype.items;
}
