// @farris/ide-devkit/esm5/lib/ideframework/gsp/panel-container.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { CompositeDisposable, Emitter } from './event-kit';
var PanelContainer = /** @class */ (function () {
    function PanelContainer(viewRegistry, location, dock) {
        if (dock === void 0) { dock = null; }
        this.viewRegistry = viewRegistry;
        this.location = location;
        this.dock = dock;
        this.panels = [];
        this.emitter = new Emitter();
        this.subscriptions = new CompositeDisposable();
    }
    /**
     * @return {?}
     */
    PanelContainer.prototype.destroy = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        try {
            for (var _b = tslib_1.__values(this.getPanels()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var panel = _c.value;
                panel.destroy();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    /**
     * @return {?}
     */
    PanelContainer.prototype.getElement = /**
     * @return {?}
     */
    function () {
    };
    /*
    Section: Event Subscription
    */
    /*
      Section: Event Subscription
      */
    /**
     * @param {?} callback
     * @return {?}
     */
    PanelContainer.prototype.onDidAddPanel = /*
      Section: Event Subscription
      */
    /**
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        return this.emitter.on('did-add-panel', callback);
    };
    /**
     * @param {?} callback
     * @return {?}
     */
    PanelContainer.prototype.onDidRemovePanel = /**
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        return this.emitter.on('did-remove-panel', callback);
    };
    /**
     * @param {?} callback
     * @return {?}
     */
    PanelContainer.prototype.onDidDestroy = /**
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        return this.emitter.once('did-destroy', callback);
    };
    /**
     * @return {?}
     */
    PanelContainer.prototype.getLocation = /**
     * @return {?}
     */
    function () {
        return this.location;
    };
    /**
     * @return {?}
     */
    PanelContainer.prototype.isModal = /**
     * @return {?}
     */
    function () { return this.location === 'modal'; };
    /**
     * @return {?}
     */
    PanelContainer.prototype.getPanels = /**
     * @return {?}
     */
    function () { return this.panels.slice(); };
    /**
     * @param {?} panel
     * @return {?}
     */
    PanelContainer.prototype.addPanel = /**
     * @param {?} panel
     * @return {?}
     */
    function (panel) {
        this.subscriptions.add(panel.onDidDestroy(this.panelDestroyed.bind(this)));
        /** @type {?} */
        var index = this.getPanelIndex(panel);
        if (index === this.panels.length) {
            this.panels.push(panel);
        }
        else {
            this.panels.splice(index, 0, panel);
        }
        this.emitter.emit('did-add-panel', { panel: panel, index: index });
        return panel;
    };
    /**
     * @param {?} item
     * @return {?}
     */
    PanelContainer.prototype.panelForItem = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        var e_2, _a;
        try {
            for (var _b = tslib_1.__values(this.panels), _c = _b.next(); !_c.done; _c = _b.next()) {
                var panel = _c.value;
                if (panel.getItem() === item) {
                    return panel;
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return null;
    };
    /**
     * @param {?} panel
     * @return {?}
     */
    PanelContainer.prototype.panelDestroyed = /**
     * @param {?} panel
     * @return {?}
     */
    function (panel) {
        /** @type {?} */
        var index = this.panels.indexOf(panel);
        if (index > -1) {
            this.panels.splice(index, 1);
            this.emitter.emit('did-remove-panel', { panel: panel, index: index });
        }
    };
    /**
     * @param {?} panel
     * @return {?}
     */
    PanelContainer.prototype.getPanelIndex = /**
     * @param {?} panel
     * @return {?}
     */
    function (panel) {
        var _this = this;
        /** @type {?} */
        var priority = panel.getPriority();
        if (['bottom', 'right'].find(function (item) { return item === _this.location; })) {
            for (var i = this.panels.length - 1; i >= 0; i--) {
                /** @type {?} */
                var p = this.panels[i];
                if (priority < p.getPriority()) {
                    return i + 1;
                }
            }
            return 0;
        }
        else {
            for (var i = 0; i < this.panels.length; i++) {
                /** @type {?} */
                var p = this.panels[i];
                if (priority < p.getPriority()) {
                    return i;
                }
            }
            return this.panels.length;
        }
    };
    return PanelContainer;
}());
export { PanelContainer };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PanelContainer.prototype.emitter;
    /**
     * @type {?}
     * @private
     */
    PanelContainer.prototype.subscriptions;
    /** @type {?} */
    PanelContainer.prototype.panels;
    /**
     * @type {?}
     * @private
     */
    PanelContainer.prototype.viewRegistry;
    /**
     * @type {?}
     * @private
     */
    PanelContainer.prototype.location;
    /**
     * @type {?}
     * @private
     */
    PanelContainer.prototype.dock;
}
