// @farris/ide-devkit/esm5/lib/ideframework/gsp/workspace.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { PanelContainer } from './panel-container';
import { Panel } from './panel';
import { Dock } from './dock';
import { WorkspaceCenter } from './workspace-center';
/**
 * @record
 */
export function PaneContainers() { }
if (false) {
    /** @type {?} */
    PaneContainers.prototype.center;
    /** @type {?} */
    PaneContainers.prototype.left;
    /** @type {?} */
    PaneContainers.prototype.right;
    /** @type {?} */
    PaneContainers.prototype.bottom;
}
/**
 * @record
 */
export function PanelContainers() { }
/** @type {?} */
var gsp = {};
/** @type {?} */
var ALL_LOCATIONS = ['center', 'left', 'right', 'bottom'];
var Workspace = /** @class */ (function () {
    function Workspace(params) {
        this.initOptions(params);
        this.initWorkspace();
        this.openers = [];
        this.frmOpeners = [];
        this.useFrmOpener = false;
        this.viewRegistry = params.viewRegistry;
        // this.subscribeToEvents();
    }
    /**
     * @private
     * @param {?} params
     * @return {?}
     */
    Workspace.prototype.initOptions = /**
     * @private
     * @param {?} params
     * @return {?}
     */
    function (params) {
        this.config = params.config || {};
    };
    /**
     * @private
     * @return {?}
     */
    Workspace.prototype.initWorkspace = /**
     * @private
     * @return {?}
     */
    function () {
        // 创建工作区停靠面板容器。
        this.paneContainers = {
            // 创建主区域。
            center: this.createCenter(),
            // 创建左侧面板。
            left: this.createDock('left'),
            // 创建右侧面板。
            right: this.createDock('right'),
            // 创建底部面板。
            bottom: this.createDock('bottom')
        };
        this.activePaneContainer = this.paneContainers.center;
        // 创建面板容器
        this.panelContainers = {
            top: new PanelContainer(this.viewRegistry, 'top'),
            left: new PanelContainer(this.viewRegistry, 'left', this.paneContainers.left),
            right: new PanelContainer(this.viewRegistry, 'right', this.paneContainers.right),
            bottom: new PanelContainer(this.viewRegistry, 'bottom', this.paneContainers.bottom),
            header: new PanelContainer(this.viewRegistry, 'header'),
            footer: new PanelContainer(this.viewRegistry, 'footer'),
            modal: new PanelContainer(this.viewRegistry, 'modal')
        };
    };
    Object.defineProperty(Workspace.prototype, "paneContainer", {
        get: /**
         * @return {?}
         */
        function () {
            return this.paneContainers.center.paneContainer;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 返回工作区元素
     */
    /**
     * 返回工作区元素
     * @return {?}
     */
    Workspace.prototype.getElement = /**
     * 返回工作区元素
     * @return {?}
     */
    function () {
    };
    /**
     * 创建主面板
     */
    /**
     * 创建主面板
     * @return {?}
     */
    Workspace.prototype.createCenter = /**
     * 创建主面板
     * @return {?}
     */
    function () {
        return new WorkspaceCenter({ viewRegistry: this.viewRegistry });
    };
    /**
     * 创建停靠面板
     * @param location 停靠面板位置
     */
    /**
     * 创建停靠面板
     * @param {?} location 停靠面板位置
     * @return {?}
     */
    Workspace.prototype.createDock = /**
     * 创建停靠面板
     * @param {?} location 停靠面板位置
     * @return {?}
     */
    function (location) {
        return new Dock({
            location: location,
            viewRegistry: this.viewRegistry
        });
    };
    /**
     * 重置工作区
     * @param packageManager 包管理器
     */
    /**
     * 重置工作区
     * @param {?} packageManager 包管理器
     * @return {?}
     */
    Workspace.prototype.reset = /**
     * 重置工作区
     * @param {?} packageManager 包管理器
     * @return {?}
     */
    function (packageManager) {
        this.packageManager = packageManager;
        this.destroyed();
        this.initWorkspace();
    };
    // Called by the Serializable mixin during serialization.
    /**
     * 序列化工作区配置
     */
    // Called by the Serializable mixin during serialization.
    /**
     * 序列化工作区配置
     * @return {?}
     */
    Workspace.prototype.serialize = 
    // Called by the Serializable mixin during serialization.
    /**
     * 序列化工作区配置
     * @return {?}
     */
    function () {
    };
    /**
     * 反序列化工作区配置
     * @param state state
     * @param deserializerManager dm
     */
    /**
     * 反序列化工作区配置
     * @param {?} state state
     * @param {?} deserializerManager dm
     * @return {?}
     */
    Workspace.prototype.deserialize = /**
     * 反序列化工作区配置
     * @param {?} state state
     * @param {?} deserializerManager dm
     * @return {?}
     */
    function (state, deserializerManager) {
    };
    // Updates the application's title and proxy icon based on whichever file is
    // open.
    // Updates the application's title and proxy icon based on whichever file is
    // open.
    /**
     * @return {?}
     */
    Workspace.prototype.updateWindowTitle = 
    // Updates the application's title and proxy icon based on whichever file is
    // open.
    /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} itemOrURI
     * @param {?=} options
     * @return {?}
     */
    Workspace.prototype.open = /**
     * @param {?} itemOrURI
     * @param {?=} options
     * @return {?}
     */
    function (itemOrURI, options) {
        if (options === void 0) { options = {}; }
        /** @type {?} */
        var uri;
        /** @type {?} */
        var item;
        if (typeof itemOrURI === 'string') {
            uri = itemOrURI; // this.project.resolvePath(itemOrURI);
        }
        else if (itemOrURI) {
            item = itemOrURI;
            if (typeof item.getURI === 'function') {
                uri = item.getURI();
            }
        }
        if (!this.config.get('core.allowPendingPaneItems')) {
            options.pending = false;
        }
        // Avoid adding URLs as recent documents to work-around this Spotlight crash:
        // https://github.com/atom/atom/issues/10071
        // if (uri && (!url.parse(uri).protocol || process.platform === 'win32')) {
        //   this.applicationDelegate.addRecentDocument(uri);
        // }
        /** @type {?} */
        var pane;
        /** @type {?} */
        var itemExistsInWorkspace;
        // Try to find an existing item in the workspace.
        if (item || uri) {
            if (options.pane) {
                pane = options.pane;
            }
            else if (options.searchAllPanes) {
                pane = item ? this.paneForItem(item) : this.paneForURI(uri);
            }
            else {
                // If an item with the given URI is already in the workspace, assume
                // that item's pane container is the preferred location for that URI.
                /** @type {?} */
                var container = void 0;
                if (uri) {
                    container = this.paneContainerForURI(uri);
                }
                if (!container) {
                    container = this.getActivePaneContainer();
                }
                // The `split` option affects where we search for the item.
                pane = container.getActivePane();
                switch (options.split) {
                    case 'left':
                        pane = pane.findLeftmostSibling();
                        break;
                    case 'right':
                        pane = pane.findRightmostSibling();
                        break;
                    case 'up':
                        pane = pane.findTopmostSibling();
                        break;
                    case 'down':
                        pane = pane.findBottommostSibling();
                        break;
                }
            }
            if (pane) {
                if (item) {
                    itemExistsInWorkspace = pane.getItems().includes(item);
                }
                else {
                    item = pane.itemForURI(uri);
                    itemExistsInWorkspace = item != null;
                }
            }
        }
        // If we already have an item at this stage, we won't need to do an async
        // lookup of the URI, so we yield the event loop to ensure this method
        // is consistently asynchronous.
        if (item) {
            Promise.resolve();
        }
        if (!itemExistsInWorkspace) {
            item = item || this.createItemForURI(uri, options);
            if (!item) {
                return;
            }
            if (options.pane) {
                pane = options.pane;
            }
            else {
                /** @type {?} */
                var location_1 = options.location;
                // if (!location && !options.split && uri && this.enablePersistence) {
                //   // location = await;
                //   this.itemLocationStore.load(uri);
                // }
                if (!location_1 && typeof item.getDefaultLocation === 'function') {
                    location_1 = item.getDefaultLocation();
                }
                /** @type {?} */
                var allowedLocations = typeof item.getAllowedLocations === 'function' ? item.getAllowedLocations() : ALL_LOCATIONS;
                location_1 = allowedLocations.includes(location_1) ? location_1 : allowedLocations[0];
                /** @type {?} */
                var container = this.paneContainers[location_1] || this.getCenter();
                pane = container.getActivePane();
                switch (options.split) {
                    case 'left':
                        pane = pane.findLeftmostSibling();
                        break;
                    case 'right':
                        pane = pane.findOrCreateRightmostSibling();
                        break;
                    case 'up':
                        pane = pane.findTopmostSibling();
                        break;
                    case 'down':
                        pane = pane.findOrCreateBottommostSibling();
                        break;
                }
            }
        }
        if (!options.pending && (pane.getPendingItem() === item)) {
            pane.clearPendingItem();
        }
        this.itemOpened(item);
        if (options.activateItem === false) {
            pane.addItem(item, { pending: options.pending });
        }
        else {
            pane.activateItem(item, { pending: options.pending });
        }
        if (options.activatePane !== false) {
            pane.activate();
        }
        /** @type {?} */
        var initialColumn = 0;
        /** @type {?} */
        var initialLine = 0;
        if (!Number.isNaN(options.initialLine)) {
            initialLine = options.initialLine;
        }
        if (!Number.isNaN(options.initialColumn)) {
            initialColumn = options.initialColumn;
        }
        if (initialLine >= 0 || initialColumn >= 0) {
            if (typeof item.setCursorBufferPosition === 'function') {
                item.setCursorBufferPosition([initialLine, initialColumn]);
            }
        }
        /** @type {?} */
        var index = pane.getActiveItemIndex();
        return item;
    };
    /**
     * @param {?} itemOrURI
     * @return {?}
     */
    Workspace.prototype.hide = /**
     * @param {?} itemOrURI
     * @return {?}
     */
    function (itemOrURI) {
        var e_1, _a, e_2, _b;
        /** @type {?} */
        var foundItems = false;
        try {
            // If any visible item has the given URI, hide it
            for (var _c = tslib_1.__values(this.getPaneContainers()), _d = _c.next(); !_d.done; _d = _c.next()) {
                var container = _d.value;
                /** @type {?} */
                var isCenter = container === this.getCenter();
                if (isCenter || ((/** @type {?} */ (container))).isVisible()) {
                    try {
                        for (var _e = tslib_1.__values(container.getPanes()), _f = _e.next(); !_f.done; _f = _e.next()) {
                            var pane = _f.value;
                            /** @type {?} */
                            var activeItem = pane.getActiveItem();
                            /** @type {?} */
                            var foundItem = (activeItem != null && (activeItem === itemOrURI ||
                                typeof activeItem.getURI === 'function' && activeItem.getURI() === itemOrURI));
                            if (foundItem) {
                                foundItems = true;
                                // We can't really hide the center so we just destroy the item.
                                if (isCenter) {
                                    pane.destroyItem(activeItem);
                                }
                                else {
                                    ((/** @type {?} */ (container))).hide();
                                }
                            }
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return foundItems;
    };
    /**
     * @param {?} itemOrURI
     * @return {?}
     */
    Workspace.prototype.toggle = /**
     * @param {?} itemOrURI
     * @return {?}
     */
    function (itemOrURI) {
        if (this.hide(itemOrURI)) {
            return Promise.resolve();
        }
        else {
            return this.open(itemOrURI, { searchAllPanes: true });
        }
    };
    /**
     * @param {?=} fileUri
     * @param {?=} options
     * @return {?}
     */
    Workspace.prototype.openSync = /**
     * @param {?=} fileUri
     * @param {?=} options
     * @return {?}
     */
    function (fileUri, options) {
        if (fileUri === void 0) { fileUri = ''; }
        if (options === void 0) { options = { initialColumn: '', initialLine: '', activateItem: '', activatePane: '' }; }
        var e_3, _a;
        var initialLine = options.initialLine, initialColumn = options.initialColumn;
        /** @type {?} */
        var activatePane = options.activatePane != null ? options.activatePane : true;
        /** @type {?} */
        var activateItem = options.activateItem != null ? options.activateItem : true;
        /** @type {?} */
        var uri = fileUri;
        // this.project.resolvePath(uri_);
        /** @type {?} */
        var item = this.getActivePane().itemForURI(uri);
        if (uri && (item == null)) {
            try {
                for (var _b = tslib_1.__values(this.getOpeners()), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var opener_1 = _c.value;
                    item = opener_1(uri, options);
                    if (item) {
                        break;
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        if (item == null) {
            item = this.project.openSync(uri, { initialLine: initialLine, initialColumn: initialColumn });
        }
        if (activateItem) {
            this.getActivePane().activateItem(item);
        }
        this.itemOpened(item);
        if (activatePane) {
            this.getActivePane().activate();
        }
        return item;
    };
    /**
     * @param {?} uri
     * @param {?} pane
     * @return {?}
     */
    Workspace.prototype.openURIInPane = /**
     * @param {?} uri
     * @param {?} pane
     * @return {?}
     */
    function (uri, pane) {
        return this.open(uri, { pane: pane });
    };
    /**
     * @param {?} uri
     * @param {?} options
     * @return {?}
     */
    Workspace.prototype.createItemForURI = /**
     * @param {?} uri
     * @param {?} options
     * @return {?}
     */
    function (uri, options) {
        var e_4, _a;
        if (uri != null) {
            try {
                for (var _b = tslib_1.__values(this.getOpeners()), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var opener_2 = _c.value;
                    /** @type {?} */
                    var item = opener_2(uri, options);
                    if (item != null) {
                        if (!item.getURI || typeof item.getURI !== 'function' || !item.getUri || typeof item.getURI !== 'function') {
                            item.getUri = function () { return uri; };
                        }
                        return item;
                    }
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
        try {
            // return this.openTextFile(uri, options);
        }
        catch (error) {
            // switch (error.code) {
            //   case 'CANCELLED':
            //     return Promise.resolve();
            //   case 'EACCES':
            //     this.notificationManager.addWarning(`Permission denied '${error.path}'`);
            //     return Promise.resolve();
            //   case 'EPERM':
            //   case 'EBUSY':
            //   case 'ENXIO':
            //   case 'EIO':
            //   case 'ENOTCONN':
            //   case 'UNKNOWN':
            //   case 'ECONNRESET':
            //   case 'EINVAL':
            //   case 'EMFILE':
            //   case 'ENOTDIR':
            //   case 'EAGAIN':
            //     this.notificationManager.addWarning(
            //       `Unable to open '${error.path != null ? error.path : uri}'`,
            //       { detail: error.message }
            //     );
            //     return Promise.resolve();
            //   default:
            //     throw error;
            // }
        }
    };
    /**
     * @return {?}
     */
    Workspace.prototype.reopenItem = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var uri = this.destroyedItemURIs.pop();
        if (uri) {
            return this.open(uri);
        }
        else {
            return Promise.resolve();
        }
    };
    /**
     * @param {?} opener
     * @return {?}
     */
    Workspace.prototype.addOpener = /**
     * @param {?} opener
     * @return {?}
     */
    function (opener) {
        this.openers.push(opener);
        // return new Disposable(() => { _.remove(this.openers, opener); });
    };
    /**
     * @param {?} opener
     * @return {?}
     */
    Workspace.prototype.addFrmOpener = /**
     * @param {?} opener
     * @return {?}
     */
    function (opener) {
        if (!this.useFrmOpener) {
            this.useFrmOpener = true;
        }
        this.frmOpeners.push(opener);
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getOpeners = /**
     * @return {?}
     */
    function () {
        return this.useFrmOpener ? this.frmOpeners : this.openers;
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getPaneItems = /**
     * @return {?}
     */
    function () {
        return new Array().concat(this.getPaneContainers().map(function (container) { return container.getPaneItems(); }));
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getActivePaneItem = /**
     * @return {?}
     */
    function () {
        return this.getActivePaneContainer().getActivePaneItem();
    };
    /**
     * @return {?}
     */
    Workspace.prototype.saveAll = /**
     * @return {?}
     */
    function () {
        this.getPaneContainers().forEach(function (container) {
            container.saveAll();
        });
    };
    /**
     * @param {?} options
     * @return {?}
     */
    Workspace.prototype.confirmClose = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return Promise.all(this.getPaneContainers().map(function (container) {
            return container.confirmClose(options);
        })).then(function (results) { return !results.find(function (item) { return item === false; }); });
    };
    /**
     * @return {?}
     */
    Workspace.prototype.saveActivePaneItem = /**
     * @return {?}
     */
    function () {
        return this.getCenter().getActivePane().saveActiveItem();
    };
    /**
     * @return {?}
     */
    Workspace.prototype.saveActivePaneItemAs = /**
     * @return {?}
     */
    function () {
        this.getCenter().getActivePane().saveActiveItemAs();
    };
    /**
     * @return {?}
     */
    Workspace.prototype.destroyActivePaneItem = /**
     * @return {?}
     */
    function () {
        return this.getActivePane().destroyActiveItem();
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getActivePaneContainer = /**
     * @return {?}
     */
    function () {
        return this.activePaneContainer;
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getPanes = /**
     * @return {?}
     */
    function () {
        return new Array().concat(this.getPaneContainers().map(function (container) { return container.getPanes(); }));
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getVisiblePanes = /**
     * @return {?}
     */
    function () {
        return new Array().concat(this.getVisiblePaneContainers().map(function (container) { return container.getPanes(); }));
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getActivePane = /**
     * @return {?}
     */
    function () {
        return this.getActivePaneContainer().getActivePane();
    };
    /**
     * @return {?}
     */
    Workspace.prototype.activateNextPane = /**
     * @return {?}
     */
    function () {
        return this.getActivePaneContainer().activateNextPane();
    };
    /**
     * @return {?}
     */
    Workspace.prototype.activatePreviousPane = /**
     * @return {?}
     */
    function () {
        return this.getActivePaneContainer().activatePreviousPane();
    };
    /**
     * @param {?} uri
     * @return {?}
     */
    Workspace.prototype.paneContainerForURI = /**
     * @param {?} uri
     * @return {?}
     */
    function (uri) {
        return this.getPaneContainers().find(function (container) { return !!container.paneForURI(uri); });
    };
    /**
     * @param {?} uri
     * @return {?}
     */
    Workspace.prototype.paneContainerForItem = /**
     * @param {?} uri
     * @return {?}
     */
    function (uri) {
        return this.getPaneContainers().find(function (container) { return !!container.paneForItem(uri); });
    };
    /**
     * @param {?} uri
     * @return {?}
     */
    Workspace.prototype.paneForURI = /**
     * @param {?} uri
     * @return {?}
     */
    function (uri) {
        var e_5, _a;
        try {
            for (var _b = tslib_1.__values(this.getPaneContainers()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var location_2 = _c.value;
                /** @type {?} */
                var pane = location_2.paneForURI(uri);
                if (pane != null) {
                    return pane;
                }
            }
        }
        catch (e_5_1) { e_5 = { error: e_5_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_5) throw e_5.error; }
        }
    };
    /**
     * @param {?} item
     * @return {?}
     */
    Workspace.prototype.paneForItem = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        var e_6, _a;
        try {
            for (var _b = tslib_1.__values(this.getPaneContainers()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var location_3 = _c.value;
                /** @type {?} */
                var pane = location_3.paneForItem(item);
                if (pane != null) {
                    return pane;
                }
            }
        }
        catch (e_6_1) { e_6 = { error: e_6_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_6) throw e_6.error; }
        }
    };
    /**
     * @return {?}
     */
    Workspace.prototype.destroyActivePane = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var activePane = this.getActivePane();
        if (activePane != null) {
            activePane.destroy();
        }
    };
    /**
     * @return {?}
     */
    Workspace.prototype.closeActivePaneItemOrEmptyPaneOrWindow = /**
     * @return {?}
     */
    function () {
        if (this.getCenter().getActivePaneItem() != null) {
            this.getCenter().getActivePane().destroyActiveItem();
        }
        else if (this.getCenter().getPanes().length > 1) {
            this.getCenter().destroyActivePane();
        }
        else if (this.config.get('core.closeEmptyWindows')) {
            // 原为关闭atom应用窗口。ide中暂不关闭整个窗口。
            // atom.close();
        }
    };
    /**
     * @param {?} item
     * @return {?}
     */
    Workspace.prototype.itemOpened = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
    };
    /**
     * @return {?}
     */
    Workspace.prototype.destroyed = /**
     * @return {?}
     */
    function () {
        this.paneContainers.center.destroy();
        this.paneContainers.left.destroy();
        this.paneContainers.right.destroy();
        this.paneContainers.bottom.destroy();
        for (var panelLocation in this.panelContainers) {
            if (this.panelContainers.hasOwnProperty(panelLocation)) {
                /** @type {?} */
                var panelContainer = this.panelContainers[panelLocation];
                this.paneContainer.destroy();
            }
        }
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getCenter = /**
     * @return {?}
     */
    function () {
        return this.paneContainers.center;
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getLeftDock = /**
     * @return {?}
     */
    function () {
        return this.paneContainers.left;
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getRightDock = /**
     * @return {?}
     */
    function () {
        return this.paneContainers.right;
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getBottomDock = /**
     * @return {?}
     */
    function () {
        return this.paneContainers.bottom;
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getPaneContainers = /**
     * @return {?}
     */
    function () {
        return [
            this.paneContainers.center,
            this.paneContainers.left,
            this.paneContainers.right,
            this.paneContainers.bottom
        ];
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getVisiblePaneContainers = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var center = this.getCenter();
        return this.getPaneContainers()
            .filter(function (container) { return container === center || ((/** @type {?} */ (container))).isVisible(); });
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getBottomPanels = /**
     * @return {?}
     */
    function () {
        return this.getPanels('bottom');
    };
    /**
     * @param {?} options
     * @return {?}
     */
    Workspace.prototype.addBottomPanel = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return this.addPanel('bottom', options);
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getLeftPanels = /**
     * @return {?}
     */
    function () {
        return this.getPanels('left');
    };
    /**
     * @param {?} options
     * @return {?}
     */
    Workspace.prototype.addLeftPanel = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return this.addPanel('left', options);
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getRightPanels = /**
     * @return {?}
     */
    function () {
        return this.getPanels('right');
    };
    /**
     * @param {?} options
     * @return {?}
     */
    Workspace.prototype.addRightPanel = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return this.addPanel('right', options);
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getTopPanels = /**
     * @return {?}
     */
    function () {
        return this.getPanels('top');
    };
    /**
     * @param {?} options
     * @return {?}
     */
    Workspace.prototype.addTopPanel = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return this.addPanel('top', options);
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getHeaderPanels = /**
     * @return {?}
     */
    function () {
        return this.getPanels('header');
    };
    /**
     * @param {?} options
     * @return {?}
     */
    Workspace.prototype.addHeaderPanel = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return this.addPanel('header', options);
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getFooterPanels = /**
     * @return {?}
     */
    function () {
        return this.getPanels('footer');
    };
    /**
     * @param {?} options
     * @return {?}
     */
    Workspace.prototype.addFooterPanel = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return this.addPanel('footer', options);
    };
    /**
     * @return {?}
     */
    Workspace.prototype.getModalPanels = /**
     * @return {?}
     */
    function () {
        return this.getPanels('modal');
    };
    /**
     * @param {?} options
     * @return {?}
     */
    Workspace.prototype.addModalPanel = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return this.addPanel('modal', options);
    };
    /**
     * @param {?} item
     * @return {?}
     */
    Workspace.prototype.panelForItem = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        for (var location_4 in this.panelContainers) {
            if (this.panelContainers.hasOwnProperty(location_4)) {
                /** @type {?} */
                var panel = this.panelContainers[location_4].panelForItem(item);
                if (panel != null) {
                    return panel;
                }
            }
        }
        return null;
    };
    /**
     * @param {?} location
     * @return {?}
     */
    Workspace.prototype.getPanelContainer = /**
     * @param {?} location
     * @return {?}
     */
    function (location) {
        return this.panelContainers[location];
    };
    /**
     * @param {?} location
     * @return {?}
     */
    Workspace.prototype.getPanels = /**
     * @param {?} location
     * @return {?}
     */
    function (location) {
        return this.panelContainers[location].getPanels();
    };
    /**
     * @param {?} location
     * @param {?} options
     * @return {?}
     */
    Workspace.prototype.addPanel = /**
     * @param {?} location
     * @param {?} options
     * @return {?}
     */
    function (location, options) {
        /** @type {?} */
        var panel = new Panel(options, this.viewRegistry);
        return this.panelContainers[location].addPanel(panel);
    };
    return Workspace;
}());
export { Workspace };
if (false) {
    /**
     * @type {?}
     * @private
     */
    Workspace.prototype.paneContainers;
    /**
     * @type {?}
     * @private
     */
    Workspace.prototype.panelContainers;
    /**
     * @type {?}
     * @private
     */
    Workspace.prototype.activePaneContainer;
    /**
     * @type {?}
     * @private
     */
    Workspace.prototype.viewRegistry;
    /**
     * @type {?}
     * @private
     */
    Workspace.prototype.packageManager;
    /**
     * @type {?}
     * @private
     */
    Workspace.prototype.openers;
    /**
     * @type {?}
     * @private
     */
    Workspace.prototype.frmOpeners;
    /**
     * @type {?}
     * @private
     */
    Workspace.prototype.useFrmOpener;
    /**
     * @type {?}
     * @private
     */
    Workspace.prototype.notificationManager;
    /**
     * @type {?}
     * @private
     */
    Workspace.prototype.config;
    /**
     * @type {?}
     * @private
     */
    Workspace.prototype.destroyedItemURIs;
    /**
     * @type {?}
     * @private
     */
    Workspace.prototype.project;
}
