// @farris/ide-devkit/esm5/lib/components/navbar/navbar.state.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
var NavbarState = /** @class */ (function () {
    function NavbarState() {
        this.dropdownInstances = [];
        this.menuItems = [];
        this.subscriptions = [];
        // 当前下拉菜单的子菜单
        this.onItemClick = new EventEmitter();
    }
    /**
     * @param {?} dropdown
     * @return {?}
     */
    NavbarState.prototype.addDropdown = /**
     * @param {?} dropdown
     * @return {?}
     */
    function (dropdown) {
        var _this = this;
        if (!this.dropdownInstances.find(function (item) { return item === dropdown; })) {
            this.dropdownInstances.push(dropdown);
            this.subscribeDropdown(dropdown);
            dropdown.onShown.subscribe(function () { return setTimeout(function () {
                dropdown.childDropdowns.forEach(function (item) { return _this.addDropdown(item); });
            }); });
        }
    };
    /**
     * @param {?} menuItem
     * @return {?}
     */
    NavbarState.prototype.addMenuItem = /**
     * @param {?} menuItem
     * @return {?}
     */
    function (menuItem) {
        if (!this.menuItems.find(function (item) { return item === menuItem; })) {
            this.menuItems.push(menuItem);
            this.subscribeMenuItem(menuItem);
        }
    };
    /**
     * @private
     * @param {?} dropdown
     * @return {?}
     */
    NavbarState.prototype.subscribeDropdown = /**
     * @private
     * @param {?} dropdown
     * @return {?}
     */
    function (dropdown) {
        // 在下拉菜单显示、隐藏时，记录下显示状态。
        // this.subscriptions.push(
        //   dropdown.onShown.subscribe((value) => {
        //     debugger;
        //     if(dropdown.isSubMenu) {
        //
        //     }
        //   })
        // );
        //
        // this.subscriptions.push(
        //   dropdown.onHidden.subscribe((value) => {
        //
        //   })
        // );
        var _this = this;
        this.subscriptions.push(dropdown.state.onClick.subscribe(function (event) {
            if (dropdown.level === 1) {
                if (dropdown.isOpen && _this.level1Dropdown === dropdown) {
                    _this.innerHideL1Dropdown();
                }
                else {
                    _this.innerShowL1Dropdown(dropdown);
                }
            }
            else if (dropdown.level === 2) {
                if (!dropdown.isOpen || _this.level2Dropdown !== dropdown) {
                    _this.innerShowL2Dropdown(dropdown);
                }
            }
            event.stopPropagation();
        }));
        this.subscriptions.push(dropdown.state.onHover.subscribe(function () {
            if (!_this.level1Dropdown) {
                // 菜单没展开，鼠标悬停没有效果
                return;
            }
            _this.show(dropdown);
        }));
    };
    /**
     * @private
     * @param {?} menuItem
     * @return {?}
     */
    NavbarState.prototype.subscribeMenuItem = /**
     * @private
     * @param {?} menuItem
     * @return {?}
     */
    function (menuItem) {
        var _this = this;
        this.subscriptions.push(menuItem.onHover.subscribe(function () {
            if (!_this.level1Dropdown) {
                return;
            }
            if (_this.level1Dropdown.containsItem(menuItem)) {
                _this.innerHideL2Dropdown();
            }
        }));
        this.subscriptions.push(menuItem.onClick.subscribe(function (command) {
            _this.onItemClick.emit(command);
        }));
    };
    /**
     * @return {?}
     */
    NavbarState.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subscriptions.forEach(function (sub) {
            sub.unsubscribe();
        });
    };
    /**
     * @return {?}
     */
    NavbarState.prototype.hideCurrentDropdown = /**
     * @return {?}
     */
    function () {
        if (this.level2Dropdown) {
            this.innerHideL2Dropdown();
        }
        else {
            this.innerHideL1Dropdown();
        }
    };
    /**
     * @return {?}
     */
    NavbarState.prototype.hideAllDropdown = /**
     * @return {?}
     */
    function () {
        this.innerHideL2Dropdown();
        this.innerHideL1Dropdown();
    };
    /**
     * @param {?} dropdown
     * @return {?}
     */
    NavbarState.prototype.show = /**
     * @param {?} dropdown
     * @return {?}
     */
    function (dropdown) {
        if (dropdown.level === 1) {
            if (dropdown !== this.level1Dropdown) {
                this.innerShowL1Dropdown(dropdown);
            }
            this.innerHideL2Dropdown();
        }
        else if (dropdown.level === 2) {
            if (dropdown !== this.level2Dropdown) {
                this.innerShowL2Dropdown(dropdown);
            }
        }
        ///////////////////////////////
        /*if (this.isMenuShown && this.isSubMenuShown && this.parentDropdown && this.parentDropdown.containsItem(dropdown)) {
          // 要显示的是子菜单，不是当前子菜单
          if (this.currentDropdown && this.currentDropdown !== dropdown) {
            this.currentDropdown.hide();
          }
          dropdown.show();
          this.currentDropdown = dropdown;
          this.isSubMenuShown = true;
        } else if (this.isMenuShown && !this.isSubMenuShown && this.currentDropdown && this.currentDropdown.containsItem(dropdown)) {
    
        } else {
          // 主菜单
          if (this.isSubMenuShown && this.currentDropdown && this.parentDropdown) {
            this.currentDropdown.hide();
            this.currentDropdown = null;
            this.isSubMenuShown = false;
          }
          if (this.isMenuShown && this.currentDropdown && this.currentDropdown !== dropdown) {
            this.currentDropdown.hide();
          }
          dropdown.show();
          this.currentDropdown = dropdown;
          this.isMenuShown = true;
        }*/
    };
    /**
     * @param {?} dropdown
     * @return {?}
     */
    NavbarState.prototype.hide = /**
     * @param {?} dropdown
     * @return {?}
     */
    function (dropdown) {
        // if
    };
    /**
     * @private
     * @param {?} target
     * @return {?}
     */
    NavbarState.prototype.innerShowL1Dropdown = /**
     * @private
     * @param {?} target
     * @return {?}
     */
    function (target) {
        if (this.level1Dropdown) {
            this.level1Dropdown.hide();
        }
        target.show();
        this.level1Dropdown = target;
    };
    /**
     * @private
     * @return {?}
     */
    NavbarState.prototype.innerHideL1Dropdown = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.level1Dropdown) {
            this.level1Dropdown.hide();
            this.level1Dropdown = null;
        }
    };
    /**
     * @private
     * @param {?} target
     * @return {?}
     */
    NavbarState.prototype.innerShowL2Dropdown = /**
     * @private
     * @param {?} target
     * @return {?}
     */
    function (target) {
        if (this.level2Dropdown) {
            this.level2Dropdown.hide();
        }
        target.show();
        this.level2Dropdown = target;
    };
    /**
     * @private
     * @return {?}
     */
    NavbarState.prototype.innerHideL2Dropdown = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.level2Dropdown) {
            this.level2Dropdown.hide();
            this.level2Dropdown = null;
        }
    };
    NavbarState.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    NavbarState.ctorParameters = function () { return []; };
    return NavbarState;
}());
export { NavbarState };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NavbarState.prototype.dropdownInstances;
    /**
     * @type {?}
     * @private
     */
    NavbarState.prototype.menuItems;
    /**
     * @type {?}
     * @private
     */
    NavbarState.prototype.subscriptions;
    /**
     * @type {?}
     * @private
     */
    NavbarState.prototype.level1Dropdown;
    /**
     * @type {?}
     * @private
     */
    NavbarState.prototype.level2Dropdown;
    /** @type {?} */
    NavbarState.prototype.onItemClick;
    /** @type {?} */
    NavbarState.prototype.autoClose;
}
