// @farris/ui-list-filter/esm5/lib/entity/conditionvalue/daterangetimevalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
var DateRangeTimeValue = /** @class */ (function () {
    function DateRangeTimeValue(value) {
        if (value === void 0) { value = { startTime: '', endTime: '' }; }
        this.clear$ = new EventEmitter();
        this.startTime = value.startTime;
        this.endTime = value.endTime;
    }
    /**
     * @return {?}
     */
    DateRangeTimeValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.startTime || !this.endTime;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    DateRangeTimeValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value.formatted) {
            this.startTime = value.formatted.split(value.delimiter)[0];
            this.endTime = value.formatted.split(value.delimiter)[1];
        }
        else {
            this.clearValue();
        }
    };
    /**
     * @return {?}
     */
    DateRangeTimeValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.startTime + "~" + this.endTime;
    };
    /**
     * @return {?}
     */
    DateRangeTimeValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.DateRangeTime;
    };
    /**
     * @return {?}
     */
    DateRangeTimeValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.startTime = undefined;
        this.endTime = undefined;
        this.clear$.emit();
    };
    return DateRangeTimeValue;
}());
export { DateRangeTimeValue };
if (false) {
    /** @type {?} */
    DateRangeTimeValue.prototype.startTime;
    /** @type {?} */
    DateRangeTimeValue.prototype.endTime;
    /** @type {?} */
    DateRangeTimeValue.prototype.clear$;
}
