// @farris/ui-list-filter/esm5/lib/entity/conditionvalue/month.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
import { EventEmitter } from "@angular/core";
var MonthValue = /** @class */ (function () {
    function MonthValue(date) {
        if (date === void 0) { date = ''; }
        this.clear$ = new EventEmitter();
        this.value = date;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    MonthValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = value.formatted;
    };
    /**
     * @return {?}
     */
    MonthValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.value;
    };
    /**
     * @return {?}
     */
    MonthValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.Date;
    };
    /**
     * @return {?}
     */
    MonthValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.value;
    };
    /**
     * @return {?}
     */
    MonthValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = undefined;
        this.clear$.emit();
    };
    return MonthValue;
}());
export { MonthValue };
if (false) {
    /** @type {?} */
    MonthValue.prototype.value;
    /** @type {?} */
    MonthValue.prototype.clear$;
}
