// @farris/ui-list-filter/esm5/lib/entity/conditionvalue/textvalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
var TextValue = /** @class */ (function () {
    function TextValue(data) {
        if (data === void 0) { data = ''; }
        this.clear$ = new EventEmitter();
        this.value = data;
    }
    /**
     * @return {?}
     */
    TextValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.Text;
    };
    /**
     * @return {?}
     */
    TextValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = undefined;
        this.clear$.emit();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TextValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        throw new Error('Method not implemented.');
    };
    /**
     * @return {?}
     */
    TextValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.value;
    };
    /**
     * @return {?}
     */
    TextValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.value;
    };
    return TextValue;
}());
export { TextValue };
if (false) {
    /** @type {?} */
    TextValue.prototype.value;
    /** @type {?} */
    TextValue.prototype.clear$;
}
