// @gsp-wf/wf-bizprocess-lookup/esm2015/lib/service/wf-bizprocess-lookup.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { map } from 'rxjs/operators';
export class WFBizProcessLookUpService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @param {?} url
     * @param {?=} params
     * @return {?}
     */
    getData(url, params) {
        /** @type {?} */
        const customData = params.customData;
        /** @type {?} */
        const index = params.pageIndex || 1;
        /** @type {?} */
        const size = params.pageSize || 20;
        /** @type {?} */
        const searchValue = JSON.parse(params.searchValue);
        if (searchValue.searchValue || customData.filter) {
            /** @type {?} */
            let query = '?param=';
            /** @type {?} */
            let filter = '{';
            if (searchValue.searchValue) {
                if (searchValue.searchField === '*') {
                    filter += `"allColumns":"${searchValue.searchValue}",`;
                }
                else {
                    filter += `"${searchValue.searchField}":"${searchValue.searchValue}",`;
                }
            }
            if (customData.filter) {
                /** @type {?} */
                const keys = Object.keys(customData.filter);
                if (keys && keys.length > 0) {
                    for (const key of keys) {
                        if (customData.filter[key]) {
                            filter += `"${key}":"${customData.filter[key]}",`;
                        }
                    }
                }
            }
            if (filter.substr(filter.length - 1, 1) === ',') {
                filter = filter.substring(0, filter.length - 1);
            }
            filter += '}';
            filter = encodeURIComponent(filter);
            query += (filter + `&pageNum=${index}&pageSize=${size}`);
            url += query;
        }
        else {
            url += `?pageNum=${index}&pageSize=${size}`;
        }
        return this.http.get(url).pipe(map((/**
         * @param {?} re
         * @return {?}
         */
        (re) => ({
            items: re.content,
            total: re.totalElements,
            pageInfo: {
                pageIndex: re.pageable.pageNumber + 1,
                pageSize: re.pageable.pageSize,
                pageList: [20, 40, 60, 80, 100],
                enablePager: true
            }
        }))));
    }
}
WFBizProcessLookUpService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
WFBizProcessLookUpService.ctorParameters = () => [
    { type: HttpService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    WFBizProcessLookUpService.prototype.http;
}
