// @farris/ui-adminorganization-selector/esm5/utils/index.type.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/index.type.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function OrgRequestParam() { }
if (false) {
    /** @type {?} */
    OrgRequestParam.prototype.absOrgType;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.layer;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.codeOrName;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.name;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.parentID;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.isSearchTree;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getCurUser;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.notGetStopOrg;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.id;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.idList;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getAllParentOrg;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getOrgAllPath;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getOrgAllPathInfo;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.withAllSub;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.layerList;
}
/** @enum {string} */
var ChangeType = {
    Append: "Append",
    Remove: "Remove",
    ValueChanged: "ValueChanged",
};
export { ChangeType };
/**
 * @record
 */
export function Param() { }
if (false) {
    /** @type {?} */
    Param.prototype.value;
    /** @type {?} */
    Param.prototype.changeType;
    /** @type {?} */
    Param.prototype.from;
}
/**
 * @record
 */
export function Organization() { }
if (false) {
    /** @type {?} */
    Organization.prototype.orgId;
    /** @type {?} */
    Organization.prototype.code;
    /** @type {?} */
    Organization.prototype.name;
    /** @type {?} */
    Organization.prototype.orgAllPath;
    /** @type {?} */
    Organization.prototype.path;
    /** @type {?} */
    Organization.prototype.layer;
    /** @type {?|undefined} */
    Organization.prototype.checked;
    /** @type {?|undefined} */
    Organization.prototype.indexId;
}
/**
 * @record
 */
export function Group() { }
if (false) {
    /** @type {?} */
    Group.prototype.public;
    /** @type {?} */
    Group.prototype.personal;
}
/**
 * @record
 */
export function GroupItem() { }
if (false) {
    /** @type {?} */
    GroupItem.prototype.id;
    /** @type {?} */
    GroupItem.prototype.title;
    /** @type {?|undefined} */
    GroupItem.prototype.index;
    /** @type {?|undefined} */
    GroupItem.prototype.description;
    /** @type {?} */
    GroupItem.prototype.total;
    /** @type {?|undefined} */
    GroupItem.prototype.avatar;
    /** @type {?|undefined} */
    GroupItem.prototype.checked;
}
/**
 * @record
 */
export function SaveGroupReqeustParam() { }
if (false) {
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.groupId;
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.groupCode;
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.title;
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.description;
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.absOrgType;
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.addIds;
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.deleteIds;
    /** @type {?} */
    SaveGroupReqeustParam.prototype.changeType;
}
/** @enum {number} */
var ValueType = {
    Value: 0,
    Express: 1,
};
export { ValueType };
ValueType[ValueType.Value] = 'Value';
ValueType[ValueType.Express] = 'Express';
/** @enum {number} */
var CompareType = {
    Equal: 0,
    NotEqual: 1,
    Greater: 2,
    GreaterOrEqual: 3,
    Less: 4,
    LessOrEqual: 5,
    Like: 6,
    LikeStartWith: 7,
    LikeEndWith: 8,
    NotLike: 9,
    NotLikeStartWith: 10,
    NotLikeEndWith: 11,
    Is: 12,
    IsNot: 13,
    In: 14,
    NotIn: 15,
};
export { CompareType };
CompareType[CompareType.Equal] = 'Equal';
CompareType[CompareType.NotEqual] = 'NotEqual';
CompareType[CompareType.Greater] = 'Greater';
CompareType[CompareType.GreaterOrEqual] = 'GreaterOrEqual';
CompareType[CompareType.Less] = 'Less';
CompareType[CompareType.LessOrEqual] = 'LessOrEqual';
CompareType[CompareType.Like] = 'Like';
CompareType[CompareType.LikeStartWith] = 'LikeStartWith';
CompareType[CompareType.LikeEndWith] = 'LikeEndWith';
CompareType[CompareType.NotLike] = 'NotLike';
CompareType[CompareType.NotLikeStartWith] = 'NotLikeStartWith';
CompareType[CompareType.NotLikeEndWith] = 'NotLikeEndWith';
CompareType[CompareType.Is] = 'Is';
CompareType[CompareType.IsNot] = 'IsNot';
CompareType[CompareType.In] = 'In';
CompareType[CompareType.NotIn] = 'NotIn';
/** @enum {number} */
var RelationType = {
    Empty: 0,
    And: 1,
    Or: 2,
};
export { RelationType };
RelationType[RelationType.Empty] = 'Empty';
RelationType[RelationType.And] = 'And';
RelationType[RelationType.Or] = 'Or';
