// @farris/ui-adminorganization-selector/esm5/utils/state.mgr.tool.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/state.mgr.tool.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Subject } from 'rxjs';
var StateMgrTool = /** @class */ (function () {
    function StateMgrTool(http) {
        this.http = http;
        this.slSubject = new Subject();
    }
    Object.defineProperty(StateMgrTool.prototype, "selectionsSubject", {
        get: /**
         * @return {?}
         */
        function () {
            return this.slSubject.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} value
     * @param {?} changeType
     * @param {?=} from
     * @return {?}
     */
    StateMgrTool.prototype.emitSelections = /**
     * @param {?} value
     * @param {?} changeType
     * @param {?=} from
     * @return {?}
     */
    function (value, changeType, from) {
        if (from === void 0) { from = 'other'; }
        this.slSubject.next({
            value: value,
            changeType: changeType,
            from: from
        });
    };
    StateMgrTool.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    StateMgrTool.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return StateMgrTool;
}());
export { StateMgrTool };
if (false) {
    /**
     * @type {?}
     * @private
     */
    StateMgrTool.prototype.slSubject;
    /**
     * @type {?}
     * @private
     */
    StateMgrTool.prototype.http;
}
