// @farris/discussion-group/esm2015/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of discussion-group
 */
export { DiscussionGroupService } from './lib/discussion-group.service';
export { DiscussionListComponent } from './lib/discussion-list/discussion-list.component';
export { DiscussionEditorComponent } from './lib/discussion-editor/discussion-editor.component';
export { MsgInfo } from './lib/discussion-editor/discussion-editor.interface';
export { DiscussionPersonnelComponent } from './lib/discussion-personnel/discussion-personnel.component';
export { DiscussionListDirective } from './lib/discussion-list/discussionList.directive';
export { DiscussionListService } from './lib/discussion-list/discussion-list.service';
export {} from './lib/discussion-list/discussionList.interface';
export { DiscussionGroupModule } from './lib/discussion-group.module';
export { SafeHtmlPipe } from './lib/pipe/safe-html.pipe';
export { AvatarHtmlPipe, GetOrgTextPipe } from './lib/pipe/controlType.pipe';
