// @farris/ui-search-box/esm5/lib/search-box.locale.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 列设置相关多语资源
 * @type {?}
 */
export var SEARCH_BOX_LOCALE = {
    en: {
        'searchbox': {
            enterKeywords: 'Please enter keywords',
            tips: 'Space can be used to represent "or" in input, and enter to search.',
            contain: 'contain',
            clear: 'Clear all conditions.',
            cancel: 'Cancel',
            ok: 'Ok',
            removeFilter: 'Delete Condition',
            anyFields: 'All',
            relation: {
                and: 'AND',
                or: 'OR'
            }
        }
    },
    'zh-CHS': {
        'searchbox': {
            enterKeywords: '请输入关键字',
            tips: '输入中可用空格代表“或”，回车可搜索',
            contain: '包含',
            clear: '清除筛选',
            cancel: '取消',
            ok: '确定',
            removeFilter: '删除条件',
            anyFields: '全部',
            relation: {
                and: '并',
                or: '或'
            }
        }
    },
    'zh-CHT': {
        searchbox: {
            enterKeywords: '請輸入關鍵字',
            tips: '輸入中可用空格代表“或”，回車可搜索',
            contain: '包含',
            clear: '清除篩選',
            cancel: '取消',
            ok: '確定',
            removeFilter: '刪除條件',
            anyFields: '全部',
            relation: {
                and: '並',
                or: '或'
            }
        }
    }
};
