// @farris/ui-search-box/esm5/lib/search-box-base.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, ElementRef, Injector, Optional, Renderer2 } from '@angular/core';
import { CommonUtils, OverLayHiddenService } from '@farris/ui-common';
import { LocaleService } from '@farris/ui-locale';
import { SEARCH_BOX_LOCALE } from './search-box.locale';
var SearchBoxBase = /** @class */ (function () {
    function SearchBoxBase(el, injector, overLayService, cd) {
        this.el = el;
        this.injector = injector;
        this.overLayService = overLayService;
        this.cd = cd;
        this.commonUtils = null;
        this.render = this.injector.get(Renderer2);
        if (!overLayService) {
            this.overLayService = new OverLayHiddenService();
        }
        this.localeSer = this.injector.get(LocaleService, null);
        this.commonUtils = this.injector.get(CommonUtils, null);
        if (this.localeSer) {
            this.localeSer.setLocaleData(SEARCH_BOX_LOCALE);
        }
    }
    /** @nocollapse */
    SearchBoxBase.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Injector },
        { type: OverLayHiddenService, decorators: [{ type: Optional }] },
        { type: ChangeDetectorRef }
    ]; };
    return SearchBoxBase;
}());
export { SearchBoxBase };
if (false) {
    /** @type {?} */
    SearchBoxBase.prototype.commonUtils;
    /** @type {?} */
    SearchBoxBase.prototype.render;
    /** @type {?} */
    SearchBoxBase.prototype.localeSer;
    /** @type {?} */
    SearchBoxBase.prototype.el;
    /**
     * @type {?}
     * @protected
     */
    SearchBoxBase.prototype.injector;
    /** @type {?} */
    SearchBoxBase.prototype.overLayService;
    /**
     * @type {?}
     * @protected
     */
    SearchBoxBase.prototype.cd;
}
