// @gsp-wf/wf-process-editor/esm2015/lib/views/aif-process-list-help/entity/RtBillProcessListResult.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-bif/common-api';
import { RtShowEntitiesWithCustomization } from './RtShowEntitiesWithCustomization';
export class RtBillProcessListResult extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.entities = new Array();
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.totalCount = 0;
        this.pageCount = 0;
        this.pageIndex = 0;
        this.pageSize = 0;
        this.entities.length = 0;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "totalCount", this.totalCount);
        BpmnModelJsonHelper.SetAttributeValue(obj, "pageCount", this.pageCount);
        BpmnModelJsonHelper.SetAttributeValue(obj, "pageIndex", this.pageIndex);
        BpmnModelJsonHelper.SetAttributeValue(obj, "pageSize", this.pageSize);
        BpmnModelJsonHelper.SetAttributeValue(obj, "entities", this.entities);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.totalCount = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "totalCount");
        this.pageCount = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "pageCount");
        this.pageIndex = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "pageIndex");
        this.pageSize = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "pageSize");
        this.entities = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "entities", RtShowEntitiesWithCustomization);
    }
}
if (false) {
    /** @type {?} */
    RtBillProcessListResult.prototype.totalCount;
    /** @type {?} */
    RtBillProcessListResult.prototype.pageCount;
    /** @type {?} */
    RtBillProcessListResult.prototype.pageIndex;
    /** @type {?} */
    RtBillProcessListResult.prototype.pageSize;
    /** @type {?} */
    RtBillProcessListResult.prototype.entities;
}
