// @gsp-wf/wf-process-editor/esm2015/lib/views/aif-process-list-help/entity/RtShowEntitiesWithCustomization.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from '@edp-bif/common-api';
import { RtShowEntity } from './RtShowEntity';
export class RtShowEntitiesWithCustomization extends RtShowEntity {
    constructor() {
        super(...arguments);
        this.customizationEntities = new Array();
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.isEnabled = false;
        this.customizationEntities.length = 0;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "customizationEntities", this.customizationEntities);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.customizationEntities = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "customizationEntities", RtShowEntitiesWithCustomization);
    }
    /**
     * @protected
     * @return {?}
     */
    getTClass() {
        return RtShowEntitiesWithCustomization;
    }
}
if (false) {
    /** @type {?} */
    RtShowEntitiesWithCustomization.prototype.isEnabled;
    /** @type {?} */
    RtShowEntitiesWithCustomization.prototype.customizationEntities;
}
