// @farris/kendo-binding/esm2015/lib/format/dateformat-placeholder.pipe.js
import { Pipe } from '@angular/core';
import { ViewModel, DateUtil } from '@farris/devkit';
export class DateFormatPlaceholderPipe {
    constructor(viewModel) {
        this.viewModel = viewModel;
    }
    transform(format, controlName) {
        if (!this.viewModel.form || !this.viewModel.form.get(controlName)) {
            return {
                year: '年',
                month: '月',
                day: '日'
            };
        }
        const value = this.viewModel.form.get(controlName).value;
        let dateStr = value;
        if (dateStr) {
            dateStr = value.toISOString();
        }
        if ((DateUtil.isEmptyDateString(dateStr) === true) && format === 'yyyy年MM月dd日') {
            return {
                year: ' ',
                month: ' ',
                day: ' '
            };
        }
        return {
            year: '年',
            month: '月',
            day: '日'
        };
    }
}
DateFormatPlaceholderPipe.decorators = [
    { type: Pipe, args: [{ name: 'dateFormatPlaceholder' },] }
];
/** @nocollapse */
DateFormatPlaceholderPipe.ctorParameters = () => [
    { type: ViewModel }
];
