// @farris/kendo-binding/esm2015/lib/format/kendogrid-filter.js
import { Directive, Input, HostListener } from '@angular/core';
import { DropDownListComponent } from '@progress/kendo-angular-dropdowns';
import { FilterService, BaseFilterCellComponent } from '@progress/kendo-angular-grid';
export class KendoGridFilterDirective extends BaseFilterCellComponent {
    constructor(dropdownList, filterService) {
        super(filterService);
        this.dropdownList = dropdownList;
        this.filterService = filterService;
    }
    get selectedValue() {
        const filter = this.filterByField(this.column.dataField);
        return filter ? filter.value : null;
    }
    get defaultItem() {
        return {
            value: null,
            name: '全部'
        };
    }
    ngOnInit() {
        this.dropdownList.valueField = 'value';
        this.dropdownList.textField = 'name';
        this.dropdownList.valuePrimitive = true;
        this.dropdownList.defaultItem = this.defaultItem;
        this.dropdownList.value = this.selectedValue;
        if (this.column.dataType === 'enum') {
            this.dropdownList.data = this.column.enumData;
        }
        else if (this.column.dataType === 'boolean') {
            this.dropdownList.data = [{ value: true, name: '是' }, { value: false, name: '否' }];
        }
    }
    onChange(value) {
        this.applyFilter(value === null ?
            this.removeFilter(this.column.dataField) :
            this.updateFilter({
                field: this.column.dataField,
                operator: 'eq',
                value: value
            }));
    }
    /*
   * 获取 单元格值  主从表支持
   * @param dataItem 行数据
   * @param field 字段
   */
    getValue(dataItem, field) {
        const index = field.indexOf('.');
        if (index < 0) {
            return dataItem[field];
        }
        const fields = field.split('.');
        // field字符串组合
        return fields.reduce((result, next) => {
            return result[next];
        }, dataItem);
    }
}
KendoGridFilterDirective.decorators = [
    { type: Directive, args: [{
                selector: '[kendogridFilter]'
            },] }
];
/** @nocollapse */
KendoGridFilterDirective.ctorParameters = () => [
    { type: DropDownListComponent },
    { type: FilterService }
];
KendoGridFilterDirective.propDecorators = {
    filter: [{ type: Input }],
    column: [{ type: Input }],
    onChange: [{ type: HostListener, args: ['valueChange', ['$event'],] }]
};
