// @farris/kendo-binding/esm2015/lib/format/treetable-format.js
import { Injectable } from '@angular/core';
export class FarrisTreetableFormatService {
    constructor() {
    }
    setFormat(treeTable) {
        const columns = treeTable.columns;
        columns.forEach((column) => {
            const dataType = column.dataType;
            if (!column.formatter) {
                if (dataType === 'date') {
                    column.formatter = {
                        type: 'datetime',
                        options: {
                            format: column.format
                        }
                    };
                }
                else if (dataType === 'number') {
                    column.formatter = {
                        type: 'number',
                        options: {
                            prefix: column.prefix ? column.prefix : '',
                            suffix: column.suffix ? column.suffix : '',
                            precision: column.precision,
                            thousand: ',',
                            stripZeros: false,
                        }
                    };
                }
                else if (dataType === 'enum') {
                    column.formatter = {
                        type: 'enum',
                        options: {
                            data: column.enumData,
                            valueField: 'value',
                            textField: 'name'
                        }
                    };
                }
                else if (dataType === 'boolean') {
                    column.formatter = {
                        type: 'enum',
                        options: {
                            data: [{ name: '是', value: true }, { name: '否', value: false }],
                            valueField: 'value',
                            textField: 'name'
                        }
                    };
                }
            }
        });
    }
}
FarrisTreetableFormatService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FarrisTreetableFormatService.ctorParameters = () => [];
