// @farris/ui-employee-selector/esm5/utils/loading-service.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/loading-service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
var DataLoadingService = /** @class */ (function () {
    /**
     * 构造函数
     * 注入@farris/ui的LoadingService
     */
    function DataLoadingService() {
        //this.loadingService = this.injector.get(LoadingService);
    }
    /**
     * 显示加载中
     */
    /**
     * 显示加载中
     * @param {?=} configOrMessage
     * @return {?}
     */
    DataLoadingService.prototype.show = /**
     * 显示加载中
     * @param {?=} configOrMessage
     * @return {?}
     */
    function (configOrMessage) {
        if (this.loadingCmp) {
            this.loadingCmp.close();
            this.loadingCmp = null;
        }
        /** @type {?} */
        var loadingConfig = this.buildLoadingConfig(configOrMessage);
        this.loadingCmp = this.loadingService.show(loadingConfig);
        return this.loadingCmp;
    };
    /**
     * 构造LoadingConfig
     * @param configOrMessage 配置对象或消息字符串
     */
    /**
     * 构造LoadingConfig
     * @private
     * @param {?} configOrMessage 配置对象或消息字符串
     * @return {?}
     */
    DataLoadingService.prototype.buildLoadingConfig = /**
     * 构造LoadingConfig
     * @private
     * @param {?} configOrMessage 配置对象或消息字符串
     * @return {?}
     */
    function (configOrMessage) {
        /** @type {?} */
        var loadingConfig;
        if (!!configOrMessage) {
            if (typeof configOrMessage === 'object') {
                loadingConfig = configOrMessage;
            }
            else {
                loadingConfig = { 'message': configOrMessage };
            }
        }
        else {
            loadingConfig = {};
        }
        if (!loadingConfig.hasOwnProperty('delay')) {
            loadingConfig.delay = 0;
        }
        return loadingConfig;
    };
    /**
     * 隐藏加载中
     */
    /**
     * 隐藏加载中
     * @return {?}
     */
    DataLoadingService.prototype.hide = /**
     * 隐藏加载中
     * @return {?}
     */
    function () {
        if (!this.loadingCmp) {
            return;
        }
        this.loadingCmp.close();
        this.loadingCmp = null;
    };
    DataLoadingService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DataLoadingService.ctorParameters = function () { return []; };
    return DataLoadingService;
}());
if (false) {
    /**
     * 加载中组件实例
     * @type {?}
     * @private
     */
    DataLoadingService.prototype.loadingCmp;
    /** @type {?} */
    DataLoadingService.prototype.loadingService;
}
export { DataLoadingService };
