// @farris/ui-employee-selector/esm5/utils/user.type.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/user.type.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function UserInfo() { }
if (false) {
    /** @type {?} */
    UserInfo.prototype.orgAllPath;
    /** @type {?} */
    UserInfo.prototype.abbreviation;
    /** @type {?} */
    UserInfo.prototype.code;
    /** @type {?} */
    UserInfo.prototype.defaultOrgId;
    /** @type {?} */
    UserInfo.prototype.defaultOrgPath;
    /** @type {?} */
    UserInfo.prototype.ext1;
    /** @type {?} */
    UserInfo.prototype.ext2;
    /** @type {?} */
    UserInfo.prototype.fullPathName;
    /** @type {?} */
    UserInfo.prototype.name;
    /** @type {?} */
    UserInfo.prototype.orgTypeId;
    /** @type {?} */
    UserInfo.prototype.secLevel;
    /** @type {?} */
    UserInfo.prototype.sortOrder;
    /** @type {?} */
    UserInfo.prototype.state;
    /** @type {?} */
    UserInfo.prototype.su;
    /** @type {?} */
    UserInfo.prototype.userId;
}
/**
 * @record
 */
export function PersonnelRequestParam() { }
if (false) {
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.pageSize;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.pageIndex;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.withAllOrg;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.notGetStopUser;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.notGetCurrentUser;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.curUser;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.getUserInfo;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.orgId;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.posId;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.orderField;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.userTypeMode;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.searchMode;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.pinyin;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.initials;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.filterId;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.groupId;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.withAllGroup;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.withPermission;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.authOp;
}
/**
 * @record
 */
export function RfRequestParam() { }
if (false) {
    /** @type {?} */
    RfRequestParam.prototype.absOrgType;
    /** @type {?} */
    RfRequestParam.prototype.bizId;
    /** @type {?} */
    RfRequestParam.prototype.objectValue;
    /** @type {?|undefined} */
    RfRequestParam.prototype.addList;
    /** @type {?|undefined} */
    RfRequestParam.prototype.deleteList;
    /** @type {?|undefined} */
    RfRequestParam.prototype.notGetStopUser;
    /** @type {?|undefined} */
    RfRequestParam.prototype.withPermission;
    /** @type {?|undefined} */
    RfRequestParam.prototype.authOp;
}
/**
 * @record
 */
export function OrgRequestParam() { }
if (false) {
    /** @type {?} */
    OrgRequestParam.prototype.absOrgType;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.layer;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.codeOrName;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.name;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.parentID;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.isSearchTree;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getCurUser;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.notGetStopOrg;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.id;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getAllParentOrg;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getOrgAllPath;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getOrgAllPathInfo;
}
/**
 * @record
 */
export function PanelMode() { }
if (false) {
    /** @type {?} */
    PanelMode.prototype.showQuickMode;
    /** @type {?} */
    PanelMode.prototype.showAdvancedMode;
}
var FilterPosition = /** @class */ (function () {
    function FilterPosition() {
    }
    return FilterPosition;
}());
export { FilterPosition };
if (false) {
    /** @type {?} */
    FilterPosition.prototype.id;
    /** @type {?} */
    FilterPosition.prototype.code;
    /** @type {?} */
    FilterPosition.prototype.name;
}
var FilterOrg = /** @class */ (function () {
    function FilterOrg() {
    }
    return FilterOrg;
}());
export { FilterOrg };
if (false) {
    /** @type {?} */
    FilterOrg.prototype.id;
    /** @type {?} */
    FilterOrg.prototype.value;
    /** @type {?} */
    FilterOrg.prototype.name;
}
/** @enum {number} */
var ValueType = {
    Value: 0,
    Express: 1,
};
export { ValueType };
ValueType[ValueType.Value] = 'Value';
ValueType[ValueType.Express] = 'Express';
/** @enum {number} */
var CompareType = {
    Equal: 0,
    NotEqual: 1,
    Greater: 2,
    GreaterOrEqual: 3,
    Less: 4,
    LessOrEqual: 5,
    Like: 6,
    LikeStartWith: 7,
    LikeEndWith: 8,
    NotLike: 9,
    NotLikeStartWith: 10,
    NotLikeEndWith: 11,
    Is: 12,
    IsNot: 13,
    In: 14,
    NotIn: 15,
};
export { CompareType };
CompareType[CompareType.Equal] = 'Equal';
CompareType[CompareType.NotEqual] = 'NotEqual';
CompareType[CompareType.Greater] = 'Greater';
CompareType[CompareType.GreaterOrEqual] = 'GreaterOrEqual';
CompareType[CompareType.Less] = 'Less';
CompareType[CompareType.LessOrEqual] = 'LessOrEqual';
CompareType[CompareType.Like] = 'Like';
CompareType[CompareType.LikeStartWith] = 'LikeStartWith';
CompareType[CompareType.LikeEndWith] = 'LikeEndWith';
CompareType[CompareType.NotLike] = 'NotLike';
CompareType[CompareType.NotLikeStartWith] = 'NotLikeStartWith';
CompareType[CompareType.NotLikeEndWith] = 'NotLikeEndWith';
CompareType[CompareType.Is] = 'Is';
CompareType[CompareType.IsNot] = 'IsNot';
CompareType[CompareType.In] = 'In';
CompareType[CompareType.NotIn] = 'NotIn';
/** @enum {number} */
var RelationType = {
    Empty: 0,
    And: 1,
    Or: 2,
};
export { RelationType };
RelationType[RelationType.Empty] = 'Empty';
RelationType[RelationType.And] = 'And';
RelationType[RelationType.Or] = 'Or';
