// @gsp-sys/rac-common/esm5/lib/component/reassign/reassign-help/reassign-help.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Output, Input, EventEmitter } from '@angular/core';
import { LookupComponent } from '@farris/ui-lookup';
import { ReAssignEntity } from '../../../common/entity';
var ReassignHelpComponent = /** @class */ (function () {
    function ReassignHelpComponent() {
        this.displayText = '';
        this.selectedKeyOld = '';
        /**
         * 导航事件
         */
        this.selectedChanged = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ReassignHelpComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    // 取数列
    // 取数列
    /**
     * @param {?} args
     * @return {?}
     */
    ReassignHelpComponent.prototype.afterKeySelected = 
    // 取数列
    /**
     * @param {?} args
     * @return {?}
     */
    function (args) {
        this.curObj = new ReAssignEntity();
        this.curKeyIdList = [];
        // 更新值
        this.curObj.id = args['id'];
        this.curObj.authorizationCode = args['authorizationCode'];
        this.curObj.authorizationName = args['authorizationName'];
        this.curObj.authfieldCode = args['authfieldCode'];
        this.curObj.authfieldName = args['authfieldName'];
        this.displayText = args['authorizationName'] + ' ' + args['authfieldName'];
        this.selectedChanged.emit(this.curObj);
        this.reassignLookup.closeDialog();
    };
    /**
     * 选中行改变后触发事件
     */
    /**
     * 选中行改变后触发事件
     * @param {?} args
     * @return {?}
     */
    ReassignHelpComponent.prototype.handleSelection = /**
     * 选中行改变后触发事件
     * @param {?} args
     * @return {?}
     */
    function (args) {
        this.curKeyId = args.node.data.id;
        this.curObj = args.node.data;
        // 首次触发
        if (this.curKeyId === '') {
            this.selectedKeyOld = this.curKeyId;
            this.selectedChanged.emit(this.curObj);
        }
        else {
            // 变化的情况下触发
            if (this.selectedKeyOld !== this.curKeyId) {
                this.selectedKeyOld = this.curKeyId;
                this.selectedChanged.emit(this.curObj);
            }
        }
    };
    ReassignHelpComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'rac-reassign-help',
                    template: "<farris-lookup #reassignLookup [displayText]=\"displayText\" style=\"width:350px\"  name=\"reassignLookup\" \r\n[showButtons]=\"true\" [title]=\"'\u8BF7\u9009\u62E9\u6743\u9650'\" [buttonsRef]=\"this.reassignHelp.btnOkAndCancleRef\">\r\n  <div style=\"padding: 10px\">\r\n    <rac-reassign-grid #reassignHelp style=\"height:450px;width:540px\" [datas]=\"datas\" calss=\"f-utils-fill\"\r\n      (selectedAfterOk)=\"afterKeySelected($event)\" (closeModalEmitter)=\"this.reassignLookup.closeDialog()\">\r\n    </rac-reassign-grid>\r\n  </div>\r\n</farris-lookup>\r\n\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ReassignHelpComponent.ctorParameters = function () { return []; };
    ReassignHelpComponent.propDecorators = {
        displayText: [{ type: Input }],
        curKeyId: [{ type: Input }],
        datas: [{ type: Input }],
        reassignLookup: [{ type: ViewChild, args: ['reassignLookup',] }],
        selectedChanged: [{ type: Output }]
    };
    return ReassignHelpComponent;
}());
export { ReassignHelpComponent };
if (false) {
    /** @type {?} */
    ReassignHelpComponent.prototype.displayText;
    /** @type {?} */
    ReassignHelpComponent.prototype.curKeyId;
    /** @type {?} */
    ReassignHelpComponent.prototype.datas;
    /** @type {?} */
    ReassignHelpComponent.prototype.curKeyIdList;
    /** @type {?} */
    ReassignHelpComponent.prototype.curObj;
    /** @type {?} */
    ReassignHelpComponent.prototype.selectedKeyOld;
    /** @type {?} */
    ReassignHelpComponent.prototype.reassignLookup;
    /**
     * 导航事件
     * @type {?}
     */
    ReassignHelpComponent.prototype.selectedChanged;
}
