// @farris/ide-devkit/esm5/lib/components/ng2-right-click-menu/sh-attach-menu.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input } from '@angular/core';
import { ShContextMenuService } from './sh-context-menu.service';
import { ShContextMenuComponent } from './sh-context-menu.component';
import { fromEvent } from 'rxjs/observable/fromEvent';
import { merge } from 'rxjs/observable/merge';
var ShAttachMenuDirective = /** @class */ (function () {
    function ShAttachMenuDirective(ctxService, elm) {
        this.ctxService = ctxService;
        this.elm = elm;
    }
    /**
     * @return {?}
     */
    ShAttachMenuDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.setupEvents();
    };
    /**
     * @private
     * @return {?}
     */
    ShAttachMenuDirective.prototype.setupEvents = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var observables = [];
        if (!this.triggers) {
            observables.push(fromEvent(this.elm.nativeElement, 'contextmenu'));
        }
        else {
            this.triggers.forEach(function (t) {
                observables.push(fromEvent(_this.elm.nativeElement, t));
            });
        }
        this.sub = merge.apply(void 0, tslib_1.__spread(observables)).subscribe(this.openMenu.bind(this));
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ShAttachMenuDirective.prototype.openMenu = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.ctxService.openMenu({
            menu: this.menu,
            mouseEvent: event,
            targetElement: this.elm,
            data: this.data
        });
    };
    /**
     * @return {?}
     */
    ShAttachMenuDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.sub.unsubscribe();
    };
    ShAttachMenuDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[shAttachMenu]'
                },] }
    ];
    /** @nocollapse */
    ShAttachMenuDirective.ctorParameters = function () { return [
        { type: ShContextMenuService },
        { type: ElementRef }
    ]; };
    ShAttachMenuDirective.propDecorators = {
        menu: [{ type: Input, args: ['shAttachMenu',] }],
        triggers: [{ type: Input, args: ['shMenuTriggers',] }],
        data: [{ type: Input, args: ['shMenuData',] }]
    };
    return ShAttachMenuDirective;
}());
export { ShAttachMenuDirective };
if (false) {
    /** @type {?} */
    ShAttachMenuDirective.prototype.menu;
    /** @type {?} */
    ShAttachMenuDirective.prototype.triggers;
    /** @type {?} */
    ShAttachMenuDirective.prototype.data;
    /** @type {?} */
    ShAttachMenuDirective.prototype.sub;
    /**
     * @type {?}
     * @private
     */
    ShAttachMenuDirective.prototype.ctxService;
    /**
     * @type {?}
     * @private
     */
    ShAttachMenuDirective.prototype.elm;
}
