// @farris/ide-devkit/esm5/lib/components/ng2-right-click-menu/sh-context-menu-item.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, Input, Optional, Output, TemplateRef } from '@angular/core';
var MenuItemContext = /** @class */ (function () {
    function MenuItemContext() {
        this.$implicit = {};
    }
    return MenuItemContext;
}());
export { MenuItemContext };
if (false) {
    /** @type {?} */
    MenuItemContext.prototype.$implicit;
}
var ShContextMenuItemDirective = /** @class */ (function () {
    function ShContextMenuItemDirective(template) {
        this.template = template;
        this.closeOnClick = true;
        this.click = new EventEmitter();
        this.context = new MenuItemContext();
    }
    /**
     * @return {?}
     */
    ShContextMenuItemDirective.prototype.setNotActive = /**
     * @return {?}
     */
    function () {
        this._active = false;
        if (this.subMenu) {
            this.subMenu.setNotActive();
        }
    };
    /**
     * @return {?}
     */
    ShContextMenuItemDirective.prototype.setActive = /**
     * @return {?}
     */
    function () {
        this._active = true;
    };
    ShContextMenuItemDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[shContextMenuItem]'
                },] }
    ];
    /** @nocollapse */
    ShContextMenuItemDirective.ctorParameters = function () { return [
        { type: TemplateRef, decorators: [{ type: Optional }] }
    ]; };
    ShContextMenuItemDirective.propDecorators = {
        subMenu: [{ type: Input }],
        divider: [{ type: Input }],
        visible: [{ type: Input }],
        disabled: [{ type: Input }],
        closeOnClick: [{ type: Input }],
        click: [{ type: Output }]
    };
    return ShContextMenuItemDirective;
}());
export { ShContextMenuItemDirective };
if (false) {
    /** @type {?} */
    ShContextMenuItemDirective.prototype.subMenu;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.divider;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.visible;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.disabled;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.closeOnClick;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.click;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.context;
    /**
     * @type {?}
     * @private
     */
    ShContextMenuItemDirective.prototype._active;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.template;
}
