// @gsp-bef/gsp-udt-metadata/fesm2015/gsp-bef-gsp-udt-metadata.js
import { GSPElementObjectType, GSPElementDataType, CefGSPAssociation, GspCommonField } from '@gsp-bef/gsp-cm-metadata';
export { GspCommonField as ɵa } from '@gsp-bef/gsp-cm-metadata';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 统一数据类型定义
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 统一数据类型定义
 */
class UnifiedDataTypeDef {
    constructor() {
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 多值业务字段
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 多值业务字段
 */
class ComplexDataTypeDef extends UnifiedDataTypeDef {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 简单数据类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 简单数据类型
 */
class SimpleDataTypeDef extends UnifiedDataTypeDef {
    constructor() {
        super(...arguments);
        /**
         * 字段数据类型(默认string)
         */
        this.DataType = GSPElementDataType.String;
        /**
         * 字段对象类型(默认None)
         */
        this.ObjectType = GSPElementObjectType.None;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 属性使用方式的信息
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 属性使用方式的信息[模板/约束，是否可改变使用方式]
 */
class UseTypeInfo {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 业务字段的json序列化结构
 */
class UdtMetadataContent {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 关联关系
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 关联关系
 */
class UdtAssociation extends CefGSPAssociation {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// import { Guid } from '../../../util/Guid';
// import { GspAssociation } from '../../ref';
// import { GSPEnumValue } from '@gsp-bef/gsp-be-metadata';
/**
 * 字段定义
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 字段定义
 */
class UdtElement extends GspCommonField {
    constructor() {
        super();
        /**
         * 字段数据类型(默认string)
         */
        this.DataType = GSPElementDataType.String;
        // /**
        //  * 字段对象类型(默认None)
        //  */
        this.ObjectType = GSPElementObjectType.None;
        // this.Id = Guid.newGuid();
        this.Length = 36;
        this.Precision = 0;
        this.DataType = GSPElementDataType.String;
        this.ObjectType = GSPElementObjectType.None;
        this.IsMultiLanguage = false;
        this.IsRequire = false;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 属性使用方式类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
const UseType = {
    /**
     * 约束
     */
    AsConstraint: 'AsConstraint',
    /**
     * 模板
     */
    AsTemplate: 'AsTemplate',
};
/** @type {?} */
const UdtTypes = [
    { text: '约束', value: UseType.AsConstraint },
    { text: '模板', value: UseType.AsTemplate }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 校验规则时机
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {number} */
const UdtTriggerTimePointType = {
    /// <summary>
    /// 不执行,用于判断时机比较结果
    /// </summary>
    None: 0,
    /// <summary>
    /// 数据更新后
    /// </summary>
    AfterModify: 1,
    /// <summary>
    /// 保存前
    /// </summary>
    BeforeSave: 2,
};
UdtTriggerTimePointType[UdtTriggerTimePointType.None] = 'None';
UdtTriggerTimePointType[UdtTriggerTimePointType.AfterModify] = 'AfterModify';
UdtTriggerTimePointType[UdtTriggerTimePointType.BeforeSave] = 'BeforeSave';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 字段映射方式
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
const ColumnMapType = {
    /**
     * 对应单列
     */
    SingleColumn: 'SingleColumn',
    /**
     * 对应多列（一一对应）
     */
    MultiColumns: 'MultiColumns',
};
/** @type {?} */
const ColumnMapTypes = [
    { text: '存为多列', value: ColumnMapType.MultiColumns },
    { text: '存为一列', value: ColumnMapType.SingleColumn }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 多值业务字段数据库配置
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 多值业务字段
 */
class UdtDbInfo {
    constructor() {
        /**
         * 字段映射方式
         */
        this.MappingType = ColumnMapType.MultiColumns;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * udt校验规则
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * udt校验规则
 */
class ValidationInfo {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { ComplexDataTypeDef, SimpleDataTypeDef, UnifiedDataTypeDef, UseTypeInfo, UdtMetadataContent, UdtAssociation, UdtElement, UseType, UdtTypes, UdtTriggerTimePointType, ColumnMapType, ColumnMapTypes, UdtDbInfo, ValidationInfo };
