// @farris/mobile-devkit/esm2015/lib/http/http-client.js
import { from, of } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import axios from 'axios';
import { HttpUtil } from './http-util';
/**
 * HttpClient
 */
class HttpClient {
    /**
     * 构造函数
     */
    constructor() {
        this.axiosInstance = axios.create();
    }
    /**
     * 发送GET请求
     */
    get(url, requestConfig) {
        return this.request('GET', url, requestConfig);
    }
    /**
     * 发送POST请求
     */
    post(url, body, requestConfig) {
        requestConfig = HttpUtil.appendBodyToRequestConfig(body, requestConfig);
        return this.request('POST', url, requestConfig);
    }
    /**
     * 发送PUT请求
     */
    put(url, body, requestConfig) {
        requestConfig = HttpUtil.appendBodyToRequestConfig(body, requestConfig);
        return this.request('PUT', url, requestConfig);
    }
    /**
     * 发送PATCH请求
     */
    patch(url, body, requestConfig) {
        requestConfig = HttpUtil.appendBodyToRequestConfig(body, requestConfig);
        return this.request('PATCH', url, requestConfig);
    }
    /**
     * 发送DELETE请求
     */
    delete(url, requestConfig) {
        return this.request('DELETE', url, requestConfig);
    }
    /**
     * 发送请求
     */
    request(method, url, requestConfig) {
        const request$ = of(true).pipe(switchMap(() => {
            const axiosRequestConfig = HttpUtil.buildAxiosRequestConfig(method, url, requestConfig);
            return from(this.axiosInstance.request(axiosRequestConfig));
        }));
        return request$.pipe(map((axiosResponse) => {
            const httpResponse = HttpUtil.buildHttpResponse(axiosResponse);
            return requestConfig.observe === 'response' ? httpResponse : axiosResponse.data;
        }));
    }
}
export { HttpClient };
