// @gsp-sys/rac-common/esm5/lib/common/service/auresult-service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { RestfulService } from '@farris/devkit';
import { RacCommonUtil } from '../../common/util/rac-util';
import { AuthDataSource } from '../entity/auth-data-source.entity';
// import { isNullOrEmptyString } from '@progress/kendo-angular-grid/dist/es2015/utils';
var AuresultService = /** @class */ (function () {
    function AuresultService(rest, session) {
        this.rest = rest;
        this.session = session;
    }
    /**
     * 获取权限需要分配的业务数据的数据源
     */
    /**
     * 获取权限需要分配的业务数据的数据源
     * @param {?} authField
     * @param {?} resultContext
     * @return {?}
     */
    AuresultService.prototype.getDataSourceByContext = /**
     * 获取权限需要分配的业务数据的数据源
     * @param {?} authField
     * @param {?} resultContext
     * @return {?}
     */
    function (authField, resultContext) {
        // mock数据
        // return of(dataSource);this.resultContext.dictDimensionValue
        /** @type {?} */
        var headers = this.session.getHeader();
        /** @type {?} */
        var url = this.session.getResultPath();
        /** @type {?} */
        var urlNew = url + '/datasource';
        /** @type {?} */
        var year = '';
        if (resultContext.yearObj && resultContext.yearObj.value) {
            year = resultContext.yearObj.value;
        }
        resultContext.year = year;
        if (resultContext.auVarValues && resultContext.auVarValues.length > 0) {
            return this.rest.post(urlNew + '/var', resultContext, null, headers);
        }
        /** @type {?} */
        var ownerId = resultContext.ownerId;
        /** @type {?} */
        var ownerType = resultContext.ownerType;
        /** @type {?} */
        var param = {
            ownerId: ownerId,
            ownerType: ownerType,
            authfield: authField,
            year: year,
        };
        return this.rest.get(urlNew, param, headers);
    };
    /**
     * 获取权限可以分配的业务数据的数据源
     */
    /**
     * 获取权限可以分配的业务数据的数据源
     * @param {?} authField
     * @return {?}
     */
    AuresultService.prototype.getDataSource = /**
     * 获取权限可以分配的业务数据的数据源
     * @param {?} authField
     * @return {?}
     */
    function (authField) {
        /** @type {?} */
        var ownerId = this.session.ownerId;
        /** @type {?} */
        var ownerType = this.session.ownerType;
        /** @type {?} */
        var authDataSource = new AuthDataSource();
        /** @type {?} */
        var headers = this.session.getHeader();
        /** @type {?} */
        var url = this.session.getResultPath();
        /**
         * ownerId  权限所有者Id,此处用户受全局业务组织过滤的，查找已分配的业务组织使用，这里不需要
         * ownerType 权限所有者类型 岗位1，用户2
         * @type {?}
         */
        var param = {
            ownerId: '',
            ownerType: 2,
            authfield: authField,
            year: ''
        };
        /** @type {?} */
        var urlNew = url + '/datasource';
        return this.rest.get(urlNew, param, headers);
    };
    /**
     * 获取权限可以分配的业务数据的数据源
     */
    /**
     * 获取权限可以分配的业务数据的数据源
     * @param {?} authField
     * @param {?} year
     * @return {?}
     */
    AuresultService.prototype.getNoUsersDataSource = /**
     * 获取权限可以分配的业务数据的数据源
     * @param {?} authField
     * @param {?} year
     * @return {?}
     */
    function (authField, year) {
        /** @type {?} */
        var ownerId = 'nouser';
        /** @type {?} */
        var ownerType = this.session.ownerType;
        /** @type {?} */
        var authDataSource = new AuthDataSource();
        /** @type {?} */
        var headers = this.session.getHeader();
        /** @type {?} */
        var url = this.session.getResultPath();
        /**
         * ownerId  权限所有者Id,此处用户受全局业务组织过滤的，查找已分配的业务组织使用，这里不需要
         * ownerType 权限所有者类型 岗位1，用户2
         * @type {?}
         */
        var param = {
            ownerId: ownerId,
            ownerType: 2,
            authfield: authField,
            year: year
        };
        /** @type {?} */
        var urlNew = url + '/datasource';
        return this.rest.get(urlNew, param, headers);
    };
    /**
     * 获取权限需要分配的业务数据的数据源
     */
    /**
     * 获取权限需要分配的业务数据的数据源
     * @param {?} authField
     * @param {?} year
     * @param {?} assid1
     * @param {?} assid2
     * @return {?}
     */
    AuresultService.prototype.getNoUsersDataSourceByGroup = /**
     * 获取权限需要分配的业务数据的数据源
     * @param {?} authField
     * @param {?} year
     * @param {?} assid1
     * @param {?} assid2
     * @return {?}
     */
    function (authField, year, assid1, assid2) {
        // mock数据
        // return of(dataSource);
        /** @type {?} */
        var authDataSource = new AuthDataSource();
        /** @type {?} */
        var headers = this.session.getHeader();
        /** @type {?} */
        var url = this.session.getResultPath();
        /** @type {?} */
        var param = {
            ownerId: 'nouser',
            ownerType: 2,
            authfield: authField,
            year: year,
            assid1: assid1,
            assid2: assid2,
        };
        /** @type {?} */
        var urlNew = url + '/datasource/group';
        return this.rest.get(urlNew, param, headers);
    };
    /**
     * 获取年度
     */
    /**
     * 获取年度
     * @param {?} su
     * @return {?}
     */
    AuresultService.prototype.getYears = /**
     * 获取年度
     * @param {?} su
     * @return {?}
     */
    function (su) {
        /** @type {?} */
        var authDataSource = new AuthDataSource();
        /** @type {?} */
        var headers = this.session.getHeader();
        /** @type {?} */
        var url = this.session.getResultPath();
        if (!su) {
            su = '*';
        }
        /** @type {?} */
        var urlNew = url + '/years/' + su;
        return this.rest.get(urlNew, null, headers);
    };
    AuresultService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AuresultService.ctorParameters = function () { return [
        { type: RestfulService },
        { type: RacCommonUtil }
    ]; };
    return AuresultService;
}());
export { AuresultService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuresultService.prototype.rest;
    /**
     * @type {?}
     * @private
     */
    AuresultService.prototype.session;
}
