// @gsp-sys/rac-common/esm5/lib/common/service/authobj.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { RestfulService } from '@farris/devkit';
import { RacCommonUtil } from '../util/rac-util';
import { of } from 'rxjs/observable/of';
import { MessagerService } from '@farris/ui-messager';
/**
 * 取数服务
 */
var AuthObjService = /** @class */ (function () {
    function AuthObjService(rest, util, msgService) {
        this.rest = rest;
        this.util = util;
        this.msgService = msgService;
    }
    /**
     * 获取指定权限对象、数据操作下关联的权限字段
     */
    /**
     * 获取指定权限对象、数据操作下关联的权限字段
     * @param {?} authObjId
     * @param {?} authOpId
     * @return {?}
     */
    AuthObjService.prototype.getAuthEntryByOp = /**
     * 获取指定权限对象、数据操作下关联的权限字段
     * @param {?} authObjId
     * @param {?} authOpId
     * @return {?}
     */
    function (authObjId, authOpId) {
        /** @type {?} */
        var url = this.util.getAuthorizationPath() + "/" + authObjId + "/" + authOpId;
        /** @type {?} */
        var headers = this.util.getHeader();
        return this.rest.get(url, null, headers);
    };
    /**
     * @return {?}
     */
    AuthObjService.prototype.getAll = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var headers = this.util.getHeader();
        /** @type {?} */
        var url = this.util.getAuthorizationPath() + '/designtime';
        return this.rest.get(url, null, headers);
    };
    /**
     * @return {?}
     */
    AuthObjService.prototype.getAllReAssign = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var headers = this.util.getHeader();
        /** @type {?} */
        var url = this.util.getAuthorizationPath() + '/reassign';
        return this.rest.get(url, null, headers);
    };
    /**
     * 异常处理
     * @param auobj 权限对象实体
     */
    /**
     * 异常处理
     * @param {?} error
     * @return {?}
     */
    AuthObjService.prototype.handlerError = /**
     * 异常处理
     * @param {?} error
     * @return {?}
     */
    function (error) {
        console.log(error);
        this.msgService.error(error.error.Message);
    };
    /**
     * @private
     * @template T
     * @param {?=} operation
     * @param {?=} result
     * @return {?}
     */
    AuthObjService.prototype.handleError = /**
     * @private
     * @template T
     * @param {?=} operation
     * @param {?=} result
     * @return {?}
     */
    function (operation, result) {
        if (operation === void 0) { operation = 'operation'; }
        return (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            console.error(error); // log to console instead
            return of((/** @type {?} */ (result)));
        });
    };
    AuthObjService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AuthObjService.ctorParameters = function () { return [
        { type: RestfulService },
        { type: RacCommonUtil },
        { type: MessagerService }
    ]; };
    return AuthObjService;
}());
export { AuthObjService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthObjService.prototype.rest;
    /**
     * @type {?}
     * @private
     */
    AuthObjService.prototype.util;
    /**
     * @type {?}
     * @private
     */
    AuthObjService.prototype.msgService;
}
