// @gsp-sys/rac-common/esm5/lib/common/service/tree.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { RacCommonUtil } from '../util/rac-util';
/**
 * 取数服务
 */
var TreeService = /** @class */ (function () {
    function TreeService(util) {
        this.util = util;
    }
    /** 按照path码构造树 */
    /**
     * 按照path码构造树
     * @param {?} datas
     * @param {?} pathField
     * @return {?}
     */
    TreeService.prototype.makeTreeByPath = /**
     * 按照path码构造树
     * @param {?} datas
     * @param {?} pathField
     * @return {?}
     */
    function (datas, pathField) {
        var _this = this;
        // 上級爲空，或者沒有上級的
        /** @type {?} */
        var pathLayerOnes = datas.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            if (_this.util.isNullOrEmptyString(x[pathField])) {
                return true;
            }
            /** @type {?} */
            var findObj = datas.find((/**
             * @param {?} parent
             * @return {?}
             */
            function (parent) { return parent[pathField] === x[pathField].substring(0, x[pathField].length - 4); }));
            if (findObj === undefined || findObj === null) {
                return true;
            }
        }));
        /** @type {?} */
        var treedata = [];
        pathLayerOnes.forEach((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            /** @type {?} */
            var entity = {};
            entity['data'] = x;
            entity['children'] = [];
            treedata.push(entity);
            // this.root = x['id'];
        }));
        treedata.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            _this.findChildrenByPath(item, datas, pathField);
        }));
        return treedata;
    };
    /** 找到子节点 */
    /**
     * 找到子节点
     * @param {?} parent
     * @param {?} datas
     * @param {?} pathField
     * @return {?}
     */
    TreeService.prototype.findChildrenByPath = /**
     * 找到子节点
     * @param {?} parent
     * @param {?} datas
     * @param {?} pathField
     * @return {?}
     */
    function (parent, datas, pathField) {
        var _this = this;
        /** @type {?} */
        var childList = datas.filter((/**
         * @param {?} childitem
         * @return {?}
         */
        function (childitem) {
            if (childitem[pathField].substring(0, childitem[pathField].length - 4) === parent['data'][pathField]) {
                return true;
            }
        }));
        if (childList && childList.length > 0) {
            childList.forEach((/**
             * @param {?} child
             * @return {?}
             */
            function (child) {
                /** @type {?} */
                var childEntity = {};
                childEntity['data'] = child;
                // tslint:disable-next-line:no-string-literal
                childEntity['children'] = [];
                // tslint:disable-next-line:no-string-literal
                parent['children'].push(childEntity);
                _this.findChildrenByPath(childEntity, datas, pathField);
            }));
        }
    };
    /** 按照parent码构造树 */
    /**
     * 按照parent码构造树
     * @param {?} datas
     * @param {?} parentField
     * @param {?} keyField
     * @return {?}
     */
    TreeService.prototype.makeTreeByParent = /**
     * 按照parent码构造树
     * @param {?} datas
     * @param {?} parentField
     * @param {?} keyField
     * @return {?}
     */
    function (datas, parentField, keyField) {
        var _this = this;
        // 上級爲空，或者沒有上級的
        /** @type {?} */
        var pathLayerOnes = datas.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            if (_this.util.isNullOrEmptyString(x[parentField])) {
                return true;
            }
            // 断层的这种，应该也属于一级
            /** @type {?} */
            var p = datas.find((/**
             * @param {?} parent
             * @return {?}
             */
            function (parent) { return parent[keyField] === x[parentField]; }));
            if (!p) {
                return true;
            }
        }));
        /** @type {?} */
        var treedata = [];
        pathLayerOnes.forEach((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            /** @type {?} */
            var entity = {};
            entity['data'] = x;
            entity['children'] = [];
            treedata.push(entity);
            // this.root = x['id'];
        }));
        treedata.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            _this.findChildrenByParent(item, datas, parentField, keyField);
        }));
        return treedata;
    };
    /** 找到子节点 */
    /**
     * 找到子节点
     * @param {?} parent
     * @param {?} datas
     * @param {?} parentField
     * @param {?} keyField
     * @return {?}
     */
    TreeService.prototype.findChildrenByParent = /**
     * 找到子节点
     * @param {?} parent
     * @param {?} datas
     * @param {?} parentField
     * @param {?} keyField
     * @return {?}
     */
    function (parent, datas, parentField, keyField) {
        var _this = this;
        /** @type {?} */
        var childList = datas.filter((/**
         * @param {?} childitem
         * @return {?}
         */
        function (childitem) {
            if (childitem[parentField] === parent['data'][keyField]) {
                return true;
            }
        }));
        if (childList && childList.length > 0) {
            childList.forEach((/**
             * @param {?} child
             * @return {?}
             */
            function (child) {
                /** @type {?} */
                var childEntity = {};
                childEntity['data'] = child;
                childEntity['children'] = [];
                parent['children'].push(childEntity);
                _this.findChildrenByParent(childEntity, datas, parentField, keyField);
            }));
        }
    };
    TreeService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    TreeService.ctorParameters = function () { return [
        { type: RacCommonUtil }
    ]; };
    return TreeService;
}());
export { TreeService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TreeService.prototype.util;
}
