// @gsp-lcm/metadata-selector/esm5/metadata/metadata-select/components/metadatapackageselect/metadatapackageselect.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/metadatapackageselect/metadatapackageselect.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, EventEmitter, Output } from '@angular/core';
import { MetadataService } from '../../metadata.service';
import { TreeclassifierComponent } from '../treeclassifier/treeclassifier.component';
import { GSPMetadataService } from '../../../metadata.service';
var MetadatapackageselectComponent = /** @class */ (function () {
    function MetadatapackageselectComponent(metadataService, gspMetaService) {
        this.metadataService = metadataService;
        this.gspMetaService = gspMetaService;
        this.closeModalEmitter = new EventEmitter();
    }
    /**
     * @return {?}
     */
    MetadatapackageselectComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    MetadatapackageselectComponent.prototype.showModal = /**
     * @return {?}
     */
    function () {
        // this.isModalShown = true;
    };
    /**
     * @return {?}
     */
    MetadatapackageselectComponent.prototype.hideModal = /**
     * @return {?}
     */
    function () {
        // this.isModalShown = false;
        this.closeModalEmitter.emit();
    };
    /**
     * @return {?}
     */
    MetadatapackageselectComponent.prototype.onHidden = /**
     * @return {?}
     */
    function () {
        // this.isModalShown = false;
        this.closeModalEmitter.emit();
    };
    /**
     * @param {?} classifier
     * @return {?}
     */
    MetadatapackageselectComponent.prototype.onSelectedChange = /**
     * @param {?} classifier
     * @return {?}
     */
    function (classifier) {
        // this.classifier = classifier;
    };
    /**
     * @return {?}
     */
    MetadatapackageselectComponent.prototype.ResolvePackage = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.closeModalEmitter.emit();
        if (this.metadataService.packages.length > 0) {
            /** @type {?} */
            var flag_1 = false;
            /** @type {?} */
            var packageName_1 = this.metadataService.packages[0];
            this.gspMetaService.GetDevRootPath().subscribe((/**
             * @param {?} root
             * @return {?}
             */
            function (root) {
                _this.metadataService.rootPath = root;
                _this.metadataService.GetMetadataPackageInfo(root + _this.metadataService.packRefPath, packageName_1).subscribe((/**
                 * @param {?} packageInfo
                 * @return {?}
                 */
                function (packageInfo) {
                    /** @type {?} */
                    var header = packageInfo.header;
                    // 检测依赖项
                    _this.metadataService.GetMetadataProjectInfo().subscribe((/**
                     * @param {?} projInfo
                     * @return {?}
                     */
                    function (projInfo) {
                        // 已有依赖关系
                        /** @type {?} */
                        var refs = projInfo.metadataPackageRefs;
                        refs.forEach((/**
                         * @param {?} item
                         * @return {?}
                         */
                        function (item) {
                            if (item.name === header.name) {
                                alert('已添加对该元数据包的依赖');
                                flag_1 = true;
                            }
                        }));
                        if (flag_1 === true) {
                            return;
                        }
                        // 更新工程引用
                        _this.metadataService.UpdateMetadataProject(header).subscribe((/**
                         * @param {?} data
                         * @return {?}
                         */
                        function (data) {
                            _this.metadataService.treeData = [];
                            _this.metadataService.Refresh();
                        }));
                        // 刷新树节点
                        if (_this.treeclassifierComponent) {
                            _this.treeclassifierComponent.files = _this.metadataService.treeData;
                        }
                    }));
                }));
            }));
        }
    };
    MetadatapackageselectComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'app-metadatapackageselect',
                    template: "<!--\u7B2C\u4E8C\u79CD\u5F39\u51FA\u65B9\u6CD5\u7684\u7EC4\u4EF6-->\r\n<!-- <div *ngIf=\"isModalShown\" [config]=\"{ show: true }\" (onHidden)=\"onHidden()\" bsModal #autoShownModal=\"bs-modal\" class=\"modal fade\" tabindex=\"-1\" role=\"dialog\" aria-hidden=\"true\" style='min-height: 800px; min-width: 900px; max-width: none'>\r\n    <div class=\"modal-dialog\" style='max-width:none; width: 615px;margin-top:0'>\r\n        <div class=\"modal-content\"> -->\r\n            <!-- <div class=\"modal-header\">\r\n                <h4 class=\"modal-title pull-left\">\u9009\u62E9\u5143\u6570\u636E\u5305</h4>\r\n                <button type=\"button\" class=\"close pull-right\" aria-label=\"Close\" (click)=\"hideModal()\">\r\n          <span aria-hidden=\"true\">&times;</span>\r\n        </button>\r\n            </div> -->\r\n            <div class=\"modal-body\">\r\n                <div class=\"templateBlock\">\r\n                    <app-metadatapackagelist></app-metadatapackagelist>\r\n                </div>\r\n            </div>\r\n            <div class=\"modal-footer\">\r\n                <button type=\"button\" class=\"btn btn-default btn-primary\" (click)=\"ResolvePackage()\">\u786E\u5B9A</button>\r\n                <button type=\"button\" class=\"btn btn-default\" (click)=\"hideModal()\">\u53D6\u6D88</button>\r\n            </div>\r\n        <!-- </div>\r\n    </div>\r\n</div> -->",
                    styles: [".templateBlock{display:flex}.infoBlock{width:100%;min-width:400px}.infoBlock label{width:50px}"]
                }] }
    ];
    /** @nocollapse */
    MetadatapackageselectComponent.ctorParameters = function () { return [
        { type: MetadataService },
        { type: GSPMetadataService }
    ]; };
    MetadatapackageselectComponent.propDecorators = {
        closeModalEmitter: [{ type: Output }],
        treeclassifierComponent: [{ type: ViewChild, args: [TreeclassifierComponent,] }]
    };
    return MetadatapackageselectComponent;
}());
export { MetadatapackageselectComponent };
if (false) {
    /** @type {?} */
    MetadatapackageselectComponent.prototype.closeModalEmitter;
    /** @type {?} */
    MetadatapackageselectComponent.prototype.treeclassifierComponent;
    /**
     * @type {?}
     * @private
     */
    MetadatapackageselectComponent.prototype.metadataService;
    /**
     * @type {?}
     * @private
     */
    MetadatapackageselectComponent.prototype.gspMetaService;
}
