// @gsp-wf/task-impl-api/esm2015/lib/action-base-impl/action-base-impl.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { Subject } from 'rxjs';
export class ActionBaseImplComponent {
    constructor() {
        /**
         * 是否为批量操作
         */
        this.isBatch = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * 任务赋值
     * @param {?} t
     * @return {?}
     */
    setTask(t) {
    }
    /**
     * 动作赋值
     * @param {?} a
     * @return {?}
     */
    setAction(a) {
    }
    /**
     * 执行操作
     * @return {?}
     */
    submit() {
        /** @type {?} */
        const sub = new Subject();
        setTimeout(() => {
            console.log('重写动作执行逻辑');
            sub.next({
                result: true,
                isClose: true
            });
        }, 500);
        return sub.asObservable();
    }
    /**
     * 批量执行操作
     * @param {?} tasks
     * @return {?}
     */
    batchSubmit(tasks) {
        /** @type {?} */
        const sub = new Subject();
        setTimeout(() => {
            console.log('重写动作执行逻辑');
            sub.next({
                result: true,
                isClose: true
            });
        }, 500);
        return sub.asObservable();
    }
}
ActionBaseImplComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-action-base-impl',
                template: "<p>\r\n  action-base-impl works!\r\n</p>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
ActionBaseImplComponent.ctorParameters = () => [];
if (false) {
    /**
     * 是否为批量操作
     * @type {?}
     */
    ActionBaseImplComponent.prototype.isBatch;
    /**
     * 提交按钮名称，支持重写
     * @type {?}
     */
    ActionBaseImplComponent.prototype.submitName;
    /**
     * 任务列表上弹框的高度
     * @type {?}
     */
    ActionBaseImplComponent.prototype.dialogHeight;
    /**
     * 任务列表上弹框的宽度
     * @type {?}
     */
    ActionBaseImplComponent.prototype.dialogWidth;
    /**
     * 动作执行来源（任务列表、任务办理界面）
     * @type {?}
     */
    ActionBaseImplComponent.prototype.origin;
    /**
     * 单据
     * @type {?}
     */
    ActionBaseImplComponent.prototype.iframe;
}
