// @qdp/ide-cmp/esm5/lib/service/joint-search/joint-search.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { JointSearchManagerService, LcpService } from '@qdp/common';
var JointSearchService = /** @class */ (function () {
    function JointSearchService(jointSearchManager, lcpService) {
        var _this = this;
        this.jointSearchManager = jointSearchManager;
        this.lcpService = lcpService;
        this.jointSearchInfo = new Subject();
        this.jointSearchManager.jointsearchInfoChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.jointSearchInfo.next(value);
        }));
    }
    /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    JointSearchService.prototype.getjointsearch = /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (id, queryId, queryRelativeUrl) {
        return this.jointSearchManager.getjointsearch(id, queryId, queryRelativeUrl);
    };
    /**
     * @param {?} jointSearchInfo
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    JointSearchService.prototype.savejointsearch = /**
     * @param {?} jointSearchInfo
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (jointSearchInfo, queryRelativeUrl) {
        return this.jointSearchManager.savejointsearch(jointSearchInfo, queryRelativeUrl);
    };
    /**
     * @param {?} formid
     * @param {?} queryID
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    JointSearchService.prototype.deletejointsearch = /**
     * @param {?} formid
     * @param {?} queryID
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (formid, queryID, queryRelativeUrl) {
        return this.jointSearchManager.deletejointsearch(formid, queryID, queryRelativeUrl);
    };
    /**
     * @return {?}
     */
    JointSearchService.prototype.getLanguageList = /**
     * @return {?}
     */
    function () {
        return this.jointSearchManager.getLanguageList();
    };
    /**
     * @param {?} params
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    JointSearchService.prototype.getCol = /**
     * @param {?} params
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (params, queryRelativeUrl) {
        return this.lcpService.getCol(params, queryRelativeUrl);
    };
    JointSearchService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    JointSearchService.ctorParameters = function () { return [
        { type: JointSearchManagerService },
        { type: LcpService }
    ]; };
    return JointSearchService;
}());
export { JointSearchService };
if (false) {
    /** @type {?} */
    JointSearchService.prototype.jointSearchInfo;
    /**
     * @type {?}
     * @private
     */
    JointSearchService.prototype.jointSearchManager;
    /**
     * @type {?}
     * @private
     */
    JointSearchService.prototype.lcpService;
}
