// @farris/ui-field-group/esm2015/lib/directive/drag.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener } from '@angular/core';
export class DragDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        this.isDown = false;
        // 记录鼠标点击事件的位置 Y
        this.totalOffsetX = 0; // 记录总偏移量 X轴
        // 记录总偏移量 X轴
        this.totalOffsetY = 0; // 记录总偏移量 Y轴
    }
    // 记录总偏移量 Y轴
    // 点击事件
    /**
     * @param {?} event
     * @return {?}
     */
    onMousedown(event) {
        this.isDown = true;
        this.disX = event.clientX;
        this.disY = event.clientY;
    }
    // 监听document移动事件事件
    /**
     * @param {?} event
     * @return {?}
     */
    onMousemove(event) {
        // 判断该元素是否被点击了。
        if (this.isDown) {
            this.el.nativeElement.style.left = this.totalOffsetX + event.clientX - this.disX + 'px';
            this.el.nativeElement.style.top = this.totalOffsetY + event.clientY - this.disY + 'px';
        }
    }
    // 监听document离开事件
    /**
     * @param {?} event
     * @return {?}
     */
    onMouseup(event) {
        // 只用当元素移动过了，离开函数体才会触发。
        if (this.isDown) {
            this.totalOffsetX += event.clientX - this.disX;
            this.totalOffsetY += event.clientY - this.disY;
            this.isDown = false;
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.el.nativeElement.style.position = 'relative';
    }
}
DragDirective.decorators = [
    { type: Directive, args: [{
                selector: '[appDrag]'
            },] }
];
/** @nocollapse */
DragDirective.ctorParameters = () => [
    { type: ElementRef }
];
DragDirective.propDecorators = {
    onMousedown: [{ type: HostListener, args: ['mousedown', ['$event'],] }],
    onMousemove: [{ type: HostListener, args: ['document:mousemove', ['$event'],] }],
    onMouseup: [{ type: HostListener, args: ['document:mouseup', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    DragDirective.prototype.isDown;
    /** @type {?} */
    DragDirective.prototype.disX;
    /** @type {?} */
    DragDirective.prototype.disY;
    /**
     * @type {?}
     * @private
     */
    DragDirective.prototype.totalOffsetX;
    /**
     * @type {?}
     * @private
     */
    DragDirective.prototype.totalOffsetY;
    /** @type {?} */
    DragDirective.prototype.el;
}
