// @farris/ui-field-group/esm2015/lib/directive/swap.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef } from '@angular/core';
import { DragDirective } from './drag.directive';
/**
 * 交换位置指令
 */
export class SwapDirective extends DragDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        super(el);
        this.el = el;
    }
    /*获取元素绝对位置的纵坐标*/
    /**
     * @param {?} e
     * @return {?}
     */
    getTop(e) {
        /** @type {?} */
        let offset = e.offsetTop;
        if (e.offsetParent) {
            offset += this.getTop(e.offsetParent);
        }
        return offset;
    }
    /*获取元素绝对位置的横坐标*/
    /**
     * @param {?} e
     * @return {?}
     */
    getLeft(e) {
        /** @type {?} */
        let offset = e.offsetLeft;
        // console.log(offset);
        if (e.offsetParent) {
            offset += this.getLeft(e.offsetParent);
        }
        return offset;
    }
    // tslint:disable-next-line:use-life-cycle-interface
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
    }
}
// tslint:disable-next-line:member-ordering
/**
 * 存放所有指令位置
 */
SwapDirective.elList = new Array(Map);
SwapDirective.decorators = [
    { type: Directive, args: [{
                selector: '[appSwap]'
            },] }
];
/** @nocollapse */
SwapDirective.ctorParameters = () => [
    { type: ElementRef }
];
if (false) {
    /**
     * 存放所有指令位置
     * @type {?}
     */
    SwapDirective.elList;
    /** @type {?} */
    SwapDirective.prototype.el;
}
