// @farris/mobile-cli/lib/config/project/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.projectOutputOption = exports.projectInputOption = void 0;
const plugin_node_resolve_1 = __importDefault(require("@rollup/plugin-node-resolve"));
const plugin_commonjs_1 = __importDefault(require("@rollup/plugin-commonjs"));
const rollup_plugin_terser_1 = require("rollup-plugin-terser");
const plugin_replace_1 = __importDefault(require("@rollup/plugin-replace"));
const plugin_typescript_1 = __importDefault(require("@rollup/plugin-typescript"));
const plugin_babel_1 = require("@rollup/plugin-babel");
const rollup_plugin_vue_1 = __importDefault(require("rollup-plugin-vue"));
const rollup_plugin_css_only_1 = __importDefault(require("rollup-plugin-css-only"));
const plugin_image_1 = __importDefault(require("@rollup/plugin-image"));
const plugin_html_1 = __importDefault(require("@rollup/plugin-html"));
const rollup_plugin_copy_1 = __importDefault(require("../../plugins/rollup-plugin-copy"));
const utils_1 = require("../../utils");
const config_1 = require("./config");
const plugins = [
    plugin_node_resolve_1.default(),
    rollup_plugin_vue_1.default({
        target: 'browser',
    }),
    rollup_plugin_css_only_1.default(config_1.cssPluginOptions),
    plugin_commonjs_1.default({ sourceMap: false }),
    plugin_typescript_1.default({ declaration: false, sourceMap: false }),
    plugin_babel_1.babel(config_1.babelPluginOptions),
    plugin_image_1.default(),
    plugin_html_1.default(config_1.htmlPluginOptions),
    plugin_replace_1.default(config_1.replacePluginOptions),
    rollup_plugin_copy_1.default(config_1.copyPluginOptions),
    utils_1.ENV === 'production' && rollup_plugin_terser_1.terser({ format: { comments: false } }),
];
exports.projectInputOption = {
    input: config_1.input,
    external: config_1.EXTERNAL,
    plugins,
    onwarn: config_1.onWarn,
};
exports.projectOutputOption = config_1.output;
exports.default = {
    input: exports.projectInputOption,
    output: exports.projectOutputOption,
};
