// @farris/mobile-devkit/esm5/lib/ui-state/ui_state.js
/*
 * @Author: Witt
 * @Date: 2018-11-17 13:38:23
 * @Last Modified by: Witt
 * @Last Modified time: 2018-11-17 13:38:50
 * @todo：临时删除原有功能，待重构
 */
import { Subject } from 'rxjs';
import { UIStateMetadataUtil } from './uistate_metadata_util';
import { ControlStateListener } from './control-state/control_state_listener';
/**
 * UI状态
 */
var UIState = /** @class */ (function () {
    function UIState() {
        this.changes = new Subject();
        this.innerData = Object.assign({});
        this._init();
    }
    UIState.prototype._init = function () {
        var construct = this.constructor;
        var uiFields = UIStateMetadataUtil.getUIFields(construct);
        var controlFields = UIStateMetadataUtil.getControlFields(construct);
        this.initializeUIField(uiFields);
        this.initializeControlField(controlFields);
    };
    //注册监听
    UIState.prototype.config = function (viewModelContext) {
        var construct = this.constructor;
        var controlFields = UIStateMetadataUtil.getControlFields(construct);
        if (controlFields && Object.keys(controlFields).length > 0) {
            this.controlStateListener = viewModelContext.injector.get(ControlStateListener);
        }
    };
    UIState.prototype.initializeUIField = function (uiFieldMetadata) {
        var _this = this;
        Object.keys(uiFieldMetadata).forEach(function (propertyName) {
            var fieldMetadata = uiFieldMetadata[propertyName];
            var uiField = fieldMetadata.stateName || propertyName;
            if (delete _this[propertyName]) {
                _this.defineProperty(propertyName, uiField);
            }
        });
    };
    UIState.prototype.initializeControlField = function (controlFieldMetadata) {
        var _this = this;
        Object.keys(controlFieldMetadata).forEach(function (propertyName) {
            var fieldMetadata = controlFieldMetadata[propertyName];
            var uiField = fieldMetadata.stateName || propertyName;
            if (delete _this[propertyName]) {
                _this.defineProperty(propertyName, uiField);
            }
        });
    };
    UIState.prototype.isExistProperty = function (propertyName) {
        if (this.innerData.hasOwnProperty(propertyName) || this.hasOwnProperty(propertyName)) {
            return true;
        }
        return false;
    };
    UIState.prototype.defineProperty = function (propertyName, field) {
        if (field === void 0) { field = null; }
        Object.defineProperty(this, propertyName, {
            get: function () {
                return field !== null ? this.innerData[field] : this.innerData[propertyName];
            },
            set: function (value) {
                // 值相同时不触发变更
                var oldValue = field !== null ? this.innerData[field] : this.innerData[propertyName];
                if (oldValue === value) {
                    return;
                }
                if (field !== null) {
                    this.innerData[field] = value;
                }
                else {
                    this.innerData[propertyName] = value;
                }
                this.changes.next({
                    field: propertyName,
                    value: value
                });
            }
        });
    };
    UIState.prototype.setPropertyValue = function (propertyName, value) {
        if (propertyName === '' || propertyName === undefined) {
            return;
        }
        if (!this.isExistProperty(propertyName)) {
            this.defineProperty(propertyName);
        }
        this[propertyName] = value;
    };
    return UIState;
}());
export { UIState };
