// @edp-bif/common-api/esm5/lib/entity/TaskCenterConfig.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/TaskCenterConfig.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl, TaskCenterContent, TaskHandlingMenuArgs } from "@edp-aif/common-api";
import { JointQueryMode } from "./impl/JointQueryMode";
var TaskCenterConfig = /** @class */ (function (_super) {
    tslib_1.__extends(TaskCenterConfig, _super);
    function TaskCenterConfig() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.jointQueryMode = JointQueryMode.APP;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    TaskCenterConfig.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.taskCenterContent = null;
        this.jointQueryMode = JointQueryMode.APP;
        this.joitQueryMenu = null;
        this.jointQueryUrl = null;
    };
    /**
     * @return {?}
     */
    TaskCenterConfig.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterContent", this.taskCenterContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryMode", this.jointQueryMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "joitQueryMenu", this.joitQueryMenu);
        BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryUrl", this.jointQueryUrl);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    TaskCenterConfig.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.taskCenterContent = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "taskCenterContent", TaskCenterContent);
        this.jointQueryMode = JointQueryMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryMode")];
        this.joitQueryMenu = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "joitQueryMenu", TaskHandlingMenuArgs);
        this.jointQueryUrl = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryUrl");
    };
    return TaskCenterConfig;
}(JsonConvertableImpl));
export { TaskCenterConfig };
if (false) {
    /** @type {?} */
    TaskCenterConfig.prototype.taskCenterContent;
    /** @type {?} */
    TaskCenterConfig.prototype.jointQueryMode;
    /** @type {?} */
    TaskCenterConfig.prototype.jointQueryUrl;
    /** @type {?} */
    TaskCenterConfig.prototype.joitQueryMenu;
}
