// @gsp-dip/data-imp-exp/esm2015/lib/model/CommonStep/CommonStep.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { IEStepType } from '../IEStepType';
import { Guid } from '../utils/Guid';
/**
 * 公共步骤实体
 */
export class CommonStep {
    constructor() {
        /**
         * 内码
         */
        this.StepID = Guid.newGuid();
        /**
         *  步骤执行顺序
         */
        this.StortOrder = 1;
        /**
         * 步骤类型
         */
        this.StepType = IEStepType.Default;
        /**
         * 源模型Id
         */
        this.SrcModelId = '';
        /**
         * 目标模型Id
         */
        this.TarModelId = '';
        /**
         *  步骤名称
         */
        this.StepName = '';
        /**
         *  是否用户自定义
         */
        this.Custom = false;
        /**
         *  用户自定义构件标识
         */
        this.CustomComponent = '';
        /**
         *  步骤前事件列表
         */
        this.DoBefore = new Array();
        /**
         *  步骤后事件列表
         */
        this.DoAfter = new Array();
    }
    /**
     * copy
     * @param {?} step
     * @return {?}
     */
    static copy(step) {
        /** @type {?} */
        const commonStep = new CommonStep();
        commonStep.StepID = step.StepID;
        commonStep.StortOrder = step.StortOrder;
        commonStep.StepType = step.StepType;
        commonStep.SrcModelId = step.SrcModelId;
        commonStep.TarModelId = step.TarModelId;
        commonStep.StepName = step.StepName;
        commonStep.Custom = step.Custom;
        commonStep.CustomComponent = step.CustomComponent;
        commonStep.DoBefore = new Array();
        for (const beforeCom of step.DoBefore) {
            commonStep.DoBefore.push(beforeCom);
        }
        commonStep.DoAfter = new Array();
        for (const afterCom of step.DoAfter) {
            commonStep.DoAfter.push(afterCom);
        }
        return commonStep;
    }
}
if (false) {
    /**
     * 内码
     * @type {?}
     */
    CommonStep.prototype.StepID;
    /**
     *  步骤执行顺序
     * @type {?}
     */
    CommonStep.prototype.StortOrder;
    /**
     * 步骤类型
     * @type {?}
     */
    CommonStep.prototype.StepType;
    /**
     * 源模型Id
     * @type {?}
     */
    CommonStep.prototype.SrcModelId;
    /**
     * 目标模型Id
     * @type {?}
     */
    CommonStep.prototype.TarModelId;
    /**
     *  步骤名称
     * @type {?}
     */
    CommonStep.prototype.StepName;
    /**
     *  是否用户自定义
     * @type {?}
     */
    CommonStep.prototype.Custom;
    /**
     *  用户自定义构件标识
     * @type {?}
     */
    CommonStep.prototype.CustomComponent;
    /**
     *  步骤前事件列表
     * @type {?}
     */
    CommonStep.prototype.DoBefore;
    /**
     *  步骤后事件列表
     * @type {?}
     */
    CommonStep.prototype.DoAfter;
}
