// @gsp-dip/data-imp-exp/esm2015/lib/model/CommonStep/ServiceComponent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ActualParameter } from './ActualParameter';
import { Guid } from '../utils/Guid';
/**
 * 公共步骤组件
 */
export class ServiceComponent {
    constructor() {
        /**
         * 构件标识
         */
        this.ID = Guid.newGuid();
        /**
         * 构件标识
         */
        this.ComponentID = '';
        /**
         * 构件编号
         */
        this.ComponentCode = '';
        /**
         * 构件名称
         */
        this.ComponentName = '';
        /**
         * 元数据路径
         */
        this.ComponentMetadataPath = '';
        /**
         *  元数据FileName
         */
        this.ComponentFileName = '';
        /**
         * 方法标识
         */
        this.MethodID = '';
        /**
         * 方法编号
         */
        this.MethodCode = '';
        /**
         * 方法名称
         */
        this.MethodName = '';
        /**
         * 方法描述
         */
        this.MethodDescription = '';
        /**
         * 参数列表
         */
        this.Parameters = new Array();
    }
    /**
     * copy
     * @param {?} component
     * @return {?}
     */
    static copy(component) {
        /** @type {?} */
        const stepComponnet = new ServiceComponent();
        stepComponnet.ID = component.ID;
        stepComponnet.ComponentID = component.ComponentID;
        stepComponnet.ComponentCode = component.ComponentCode;
        stepComponnet.ComponentName = component.ComponentName;
        stepComponnet.ComponentMetadataPath = component.ComponentMetadataPath;
        stepComponnet.ComponentFileName = component.ComponentFileName;
        stepComponnet.MethodID = component.MethodID;
        stepComponnet.MethodCode = component.MethodCode;
        stepComponnet.MethodName = component.MethodName;
        stepComponnet.MethodDescription = component.MethodDescription;
        stepComponnet.Parameters = new Array();
        for (const iterator of component.Parameters) {
            stepComponnet.Parameters.push(ActualParameter.copy(iterator));
        }
        return stepComponnet;
    }
}
if (false) {
    /**
     * 构件标识
     * @type {?}
     */
    ServiceComponent.prototype.ID;
    /**
     * 构件标识
     * @type {?}
     */
    ServiceComponent.prototype.ComponentID;
    /**
     * 构件编号
     * @type {?}
     */
    ServiceComponent.prototype.ComponentCode;
    /**
     * 构件名称
     * @type {?}
     */
    ServiceComponent.prototype.ComponentName;
    /**
     * 元数据路径
     * @type {?}
     */
    ServiceComponent.prototype.ComponentMetadataPath;
    /**
     *  元数据FileName
     * @type {?}
     */
    ServiceComponent.prototype.ComponentFileName;
    /**
     * 方法标识
     * @type {?}
     */
    ServiceComponent.prototype.MethodID;
    /**
     * 方法编号
     * @type {?}
     */
    ServiceComponent.prototype.MethodCode;
    /**
     * 方法名称
     * @type {?}
     */
    ServiceComponent.prototype.MethodName;
    /**
     * 方法描述
     * @type {?}
     */
    ServiceComponent.prototype.MethodDescription;
    /**
     * 参数列表
     * @type {?}
     */
    ServiceComponent.prototype.Parameters;
}
