// @farris/ui-html-editor/esm5/lib/tool-tip-module/tooltip-list.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var tooltips = {
    "en": {
        align: {
            '': 'Align',
            center: 'Alinear al centro',
            right: 'Alinear a la derecha',
            justify: 'Justificar'
        },
        background: 'Background color',
        blockquote: 'Bloque de cita',
        bold: 'Bold',
        clean: 'Limpiar formato',
        code: 'Code',
        'code-block': 'Code block',
        color: 'Font Color',
        direction: 'Direction',
        float: {
            center: 'Center',
            full: '全浮动',
            left: '左浮动',
            right: '右浮动'
        },
        formula: 'Formula',
        header: {
            '': 'Encabezados',
            '1': 'Encabezados 1',
            '2': 'Encabezados 2'
        },
        italic: 'Italic',
        image: 'Image',
        indent: {
            '+1': 'Increase indent',
            '-1': 'Decrease indent'
        },
        link: 'Enlace',
        list: {
            ordered: 'Numbered list',
            bullet: 'Bullet list',
            check: '选择列表'
        },
        script: {
            sub: 'Subscript',
            super: 'Superscript'
        },
        strike: 'Strike',
        underline: 'Underline',
        video: 'Video',
        undo: 'Undo',
        redo: 'Redo',
        size: 'Font Size',
        font: 'Font Family',
        divider: 'Divider',
        formatBrush: 'Format Brush',
        emoji: 'Emoji'
    },
    "zh-CHS": {
        align: {
            '': '对齐',
            center: '居中对齐',
            right: '右对齐',
            justify: '两端对齐'
        },
        background: '背景色',
        blockquote: '引用',
        bold: '加粗',
        clean: '清除格式',
        code: '代码',
        'code-block': '代码段',
        color: '颜色',
        direction: '方向',
        float: {
            center: '居中',
            full: '全浮动',
            left: '左浮动',
            right: '右浮动'
        },
        formula: '公式',
        header: {
            '': '标题',
            '1': '标题1',
            '2': '标题2'
        },
        italic: '斜体',
        image: '图片',
        indent: {
            '+1': '缩进',
            '-1': '取消缩进'
        },
        link: '链接',
        list: {
            ordered: '有序列表',
            bullet: '无序列表',
            check: '选择列表'
        },
        script: {
            sub: '下标',
            super: '上标'
        },
        strike: '删除线',
        underline: '下划线',
        video: '视频',
        undo: '撤销',
        redo: '重做',
        size: '字号',
        font: '字体',
        divider: '分割线',
        formatBrush: '格式刷',
        emoji: '表情'
    },
    "zh-CHT": {
        align: {
            '': '對齊',
            center: '居中對齊',
            right: '右對齊',
            justify: '兩端對齊'
        },
        background: '背景色',
        blockquote: '引用',
        bold: '加粗',
        clean: '清除格式',
        code: '代碼',
        'code-block': '代碼段',
        color: '顔色',
        direction: '方向',
        float: {
            center: '居中',
            full: '全浮動',
            left: '左浮動',
            right: '右浮動'
        },
        formula: '公式',
        header: {
            '': '標題',
            '1': '標題1',
            '2': '標題2'
        },
        italic: '斜體',
        image: '圖片',
        indent: {
            '+1': '縮進',
            '-1': '取消縮進'
        },
        link: '鏈接',
        list: {
            ordered: '有序列表',
            bullet: '無序列表',
            check: '選擇列表'
        },
        script: {
            sub: '下標',
            super: '上標'
        },
        strike: '刪除線',
        underline: '下劃線',
        video: '視頻',
        undo: '撤銷',
        redo: '重做',
        size: '字號',
        font: '字體',
        divider: '分割線',
        formatBrush: '格式刷',
        emoji: '表情'
    }
};
