// @edp-bif/runtime-api/esm5/lib/bill-tracker/bill-tracker.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/bill-tracker.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, LOCALE_ID, NgModuleFactoryLoader, SystemJsNgModuleLoader } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { FrameworkService, FrmEventBus, FuncsService, LoadingService, AppService } from '@gsp-sys/rtf-common';
import { CacheService, SessionService, HttpService } from '@ecp-caf/caf-common';
import { AngularDraggableModule } from '@farris/ui-draggable';
import { MessagerModule } from '@farris/ui-messager';
import { LoadingModule } from '@farris/ui-loading';
import { TextModule } from '@farris/ui-text';
import { FarrisSectionModule } from '@farris/ui-section';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { DatagridModule } from '@farris/ui-datagrid';
import { BillTrackerComponent } from './components/bill-tracker/bill-tracker.component';
import { BillTrackerGraphComponent } from './components/bill-tracker-graph/bill-tracker-graph.component';
import { BillTrackerDataComponent } from './components/bill-tracker-data/bill-tracker-data.component';
import { BasicTableComponent } from './components/basic-table/basic-table.component';
import { ChildRowTableDirective } from './components/basic-table/child-row-table.directive';
import { BasicInfoComponent } from './components/basic-info/basic-info.component';
import { ChildInfoComponent } from './components/child-info/child-info.component';
import { BillTrackerDataService } from './services/bill-tracker-data.service';
import { BillTrackerUiService } from './services/bill-tracker-ui.service';
import { BillTrackerModuleConfig } from './config/bill-tracker-module.config';
import { BillTrackerLocalePipe } from './pipe/bill-tracker-locale.pipe';
import { BillTrackerDataInfoService } from './services/bill-tracker-data-info.service';
import { ChildInfoListComponent } from './components/child-info-list/child-info-list.component';
import { NotifyModule } from '@farris/ui-notify';
import { FormsModule } from '@angular/forms';
import { BillTrackerFieldLabelPipe } from './pipe/bill-tracker-field-label.pipe';
import { BillTrackerInternalUiService } from './services/bill-tracker-internal-ui.service';
var ɵ0 = localStorage.getItem('languageCode') || 'zh-CHS', ɵ1 = {};
var BillTrackerModule = /** @class */ (function () {
    function BillTrackerModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    BillTrackerModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: BillTrackerModule,
            providers: [
                { provide: BillTrackerModuleConfig, useValue: config || {} },
            ]
        };
    };
    BillTrackerModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        BillTrackerComponent,
                        BillTrackerGraphComponent,
                        BillTrackerDataComponent,
                        BasicTableComponent,
                        ChildRowTableDirective,
                        BillTrackerLocalePipe,
                        BasicInfoComponent,
                        ChildInfoComponent,
                        ChildInfoListComponent,
                        BillTrackerFieldLabelPipe,
                    ],
                    imports: [
                        CommonModule,
                        HttpClientModule,
                        FormsModule,
                        AngularDraggableModule,
                        MessagerModule.forRoot(),
                        LoadingModule.forRoot(),
                        NotifyModule,
                        TextModule,
                        FarrisSectionModule,
                        DatagridModule.forRoot(),
                        FarrisDialogModule,
                    ],
                    providers: [
                        { provide: LOCALE_ID, useValue: ɵ0 },
                        { provide: BillTrackerModuleConfig, useValue: ɵ1 },
                        { provide: NgModuleFactoryLoader, useClass: SystemJsNgModuleLoader },
                        CacheService,
                        SessionService,
                        HttpService,
                        FrmEventBus,
                        FuncsService,
                        LoadingService,
                        AppService,
                        FrameworkService,
                        BillTrackerDataService,
                        BillTrackerUiService,
                        BillTrackerInternalUiService,
                        BillTrackerLocalePipe,
                        BillTrackerDataInfoService,
                    ],
                    bootstrap: [BillTrackerComponent],
                    exports: [
                        BillTrackerComponent,
                        BillTrackerDataComponent,
                    ],
                    entryComponents: [
                        BillTrackerComponent,
                        BillTrackerDataComponent,
                        BasicTableComponent,
                        BasicInfoComponent,
                        ChildInfoComponent,
                        ChildInfoListComponent,
                    ]
                },] }
    ];
    return BillTrackerModule;
}());
export { BillTrackerModule };
export { ɵ0, ɵ1 };
