// @gsp-lcm/metadatart-selector/esm2015/metadata/metadatart-select/components/selectmetadatart/selectmetadatart.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadatart-select/components/selectmetadatart/selectmetadatart.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// tslint:disable-next-line:max-line-length
import { Component, Input, Output, EventEmitter, ComponentFactoryResolver, Injector } from '@angular/core';
import { MetadataRTService } from '../../metadatart.service';
import { Metadata4RefDto, MetadataPackageHeader } from '../../../metadata4refdto';
import { BsModalService } from '@farris/ui-modal';
import { NotifyService } from '@farris/ui-notify';
export class SelectMetadataRTComponent {
    /**
     * @param {?} metadataService
     * @param {?} modalService
     * @param {?} resolver
     * @param {?} injector
     */
    constructor(metadataService, modalService, resolver, injector) {
        this.metadataService = metadataService;
        this.modalService = modalService;
        this.resolver = resolver;
        this.injector = injector;
        this.selectedMetadata = new EventEmitter();
        this.confirm = '确定';
        this.cancel = '取消';
        this.closeModalEmitter = new EventEmitter();
        this.notifyService = this.injector.get(NotifyService, null);
        if (localStorage.getItem('languageCode') === 'en') {
            this.confirm = 'OK';
            this.cancel = 'Cancel';
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
    }
    /**
     * @return {?}
     */
    LoadMetadata() {
        if (this.metadataService.metadatas == null || this.metadataService.metadatas.length <= 0) {
            if (this.notifyService) {
                this.notifyService.warning("请先选中元数据。");
                return;
            }
            else {
                console.log("请先选中元数据。");
                return;
            }
        }
        /** @type {?} */
        const Id = this.metadataService.metadatas[0].ID;
        this.metadataService.GetMetadataRT(Id).subscribe((/**
         * @param {?} templates
         * @return {?}
         */
        templates => {
            /** @type {?} */
            const metadata4Ref = new Metadata4RefDto();
            metadata4Ref.metadata = templates;
            metadata4Ref.packageHeader = new MetadataPackageHeader();
            this.selectedMetadata.emit(metadata4Ref);
            this.hideModal();
        }));
        // 清空缓存
        this.metadataService.clearMetadataRT();
    }
    /**
     * @return {?}
     */
    hideModal() {
        this.closeModalEmitter.emit();
    }
    /**
     * @param {?} classifier
     * @return {?}
     */
    onSelectedChange(classifier) {
        this.classifier = classifier;
    }
}
SelectMetadataRTComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'app-select-metadatart',
                template: "<div class=\"farris-main-area flex-column farris-overflow-hidden\">\r\n    <div class=\"d-flex flex-fill flex-row farris-overflow-hidden\">\r\n        <app-metadatalist class=\"d-flex flex-fill\" style=\"width: 100%;\" [typeList]=\"typeList\" [boId]=\"boId\" [bizTypeId]=\"bizTypeId\"></app-metadatalist>\r\n    </div>\r\n\r\n    <div class=\"modal-footer\">\r\n        <button type=\"button\" class=\"btn btn-primary\" (click)=\"LoadMetadata()\">{{confirm}}</button>\r\n        <button type=\"button\" class=\"btn btn-default\" (click)=\"hideModal()\">{{cancel}}</button>\r\n    </div>\r\n</div>",
                styles: [".templateBlock{display:flex}.infoBlock{width:100%;min-width:400px}.infoBlock label{width:50px}"]
            }] }
];
/** @nocollapse */
SelectMetadataRTComponent.ctorParameters = () => [
    { type: MetadataRTService },
    { type: BsModalService },
    { type: ComponentFactoryResolver },
    { type: Injector }
];
SelectMetadataRTComponent.propDecorators = {
    typeList: [{ type: Input }],
    boId: [{ type: Input }],
    bizTypeId: [{ type: Input }],
    closeModalEmitter: [{ type: Output }],
    selectedMetadata: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    SelectMetadataRTComponent.prototype.typeList;
    /** @type {?} */
    SelectMetadataRTComponent.prototype.boId;
    /** @type {?} */
    SelectMetadataRTComponent.prototype.bizTypeId;
    /** @type {?} */
    SelectMetadataRTComponent.prototype.classifier;
    /** @type {?} */
    SelectMetadataRTComponent.prototype.closeModalEmitter;
    /** @type {?} */
    SelectMetadataRTComponent.prototype.selectedMetadata;
    /** @type {?} */
    SelectMetadataRTComponent.prototype.confirm;
    /** @type {?} */
    SelectMetadataRTComponent.prototype.cancel;
    /**
     * @type {?}
     * @private
     */
    SelectMetadataRTComponent.prototype.notifyService;
    /**
     * @type {?}
     * @private
     */
    SelectMetadataRTComponent.prototype.metadataService;
    /**
     * @type {?}
     * @private
     */
    SelectMetadataRTComponent.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    SelectMetadataRTComponent.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    SelectMetadataRTComponent.prototype.injector;
}
