// @gsp-wf/wf-task-handler/esm5/lib/components/assignee-info/assignee-info.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ChangeDetectorRef, Injector } from '@angular/core';
import { AssigneeInfo } from '../../entity/complete-response';
import { I18nService } from '../../service/i18n/i18n.service';
import { NotifyService } from '@farris/ui-notify';
import { TaskService } from '../../service/task.service';
import { WfContext } from '../../entity/wfContext';
import { UserHelpParam } from '../../entity/farris-user-help-param';
var AssigneeInfoComponent = /** @class */ (function () {
    function AssigneeInfoComponent(i18nService, notify, cd, taskService, injector) {
        var _this = this;
        this.i18nService = i18nService;
        this.notify = notify;
        this.cd = cd;
        this.taskService = taskService;
        this.injector = injector;
        //已经选中的指派人
        this.assignees = []; //返回选中人的ID数组
        //返回选中人的ID数组
        this.anonymous = false; // 是否匿名
        // 是否匿名
        this.anonymousText = ''; // 匿名显示文本
        //默认办理人（界面提示用）
        this.pscChange = null; //farris用户帮助
        //farris用户帮助
        this.disabled = false; //办理人员输入框是否禁用
        this.filter = { notGetCurrentUser: false, notGetStopUser: true, orderField: "order", withAllOrg: true, orgId: '' };
        this.ifShowBlank = true; //办理人为空时，是否显示选人框
        //办理人为空时，是否显示选人框
        this.openFromIframe = false; //是否从iframe打开，应用IDP提交审批集成
        //区分是哪个组件
        this.isUseNewSelectUserHelp = false; //是否直接启用新版选人组件
        //是否直接启用新版选人组件
        this.ifHorizontal = false; //是否水平排列
        //是否水平排列
        this.afterConfirm = new EventEmitter();
        this.groupIcon = '<i class="f-icon f-icon-plus-circle text-primary"></i>';
        this.extendUserHelp = false; //用户帮助是否扩展
        //是否使用farris用户帮助
        this.taskService.getWfUserHelpConfig().subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            if (re && re.userHelpConfig && re.userHelpConfig.isExtend == 'true') {
                _this.extendUserHelp = true;
                _this.extendUserHelpUrl = re.userHelpConfig.url + '?v=20221114';
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            console.error(error);
        }));
    }
    /**
     * @return {?}
     */
    AssigneeInfoComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        //引入farris用户帮助脚本
        if (this.openFromIframe) {
            ((/** @type {?} */ (top))).System.import("/platform/common/web/@farris/common-personnel-select/psc.js" + '?v=' + new Date().getTime().toString()).then((/**
             * @param {?} e
             * @return {?}
             */
            function (e) {
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) {
                console.error(error);
            }));
        }
    };
    Object.defineProperty(AssigneeInfoComponent.prototype, "assigneeInfo", {
        set: /**
         * @param {?} info
         * @return {?}
         */
        function (info) {
            this.disabled = false;
            this.placeholder = null; //清空
            this.assignUserNgModule = null; //清空
            this._assigneeInfo = info; //可以传null
            if (!this._assigneeInfo) {
                this.disabled = true;
            }
            else {
                if (this._assigneeInfo.selectAssignee) {
                    this.disabled = false;
                    this.placeholder = this.i18nService.getValue('task.selectPlaceholder'); //请选择
                }
                else {
                    this.disabled = true;
                    if (this._assigneeInfo.assigneeUsers && this._assigneeInfo.assigneeUsers.length) {
                        this.defaultParticipant = this.anonymous ? this.anonymousText : this._assigneeInfo.assigneeUsers.map((/**
                         * @param {?} user
                         * @return {?}
                         */
                        function (user) { return user.name; })).join(); //解析人名
                    }
                    else {
                        this.defaultParticipant = this.i18nService.getValue('task.noUser'); //无
                    }
                }
                this.singleSelect = !this._assigneeInfo.multiSelect;
                this.allowGlobalAssign = this._assigneeInfo.allowGlobalAssign;
                if (this._assigneeInfo.organizationScope && this._assigneeInfo.organizationScope.length) {
                    /** @type {?} */
                    var orgList = this._assigneeInfo.organizationScope[0];
                    this.filter.orgId = orgList.id;
                    this.filter.withAllOrg = orgList.includeChildren;
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AssigneeInfoComponent.prototype, "setAnonymous", {
        set: /**
         * @param {?} anonymous
         * @return {?}
         */
        function (anonymous) {
            this.anonymous = anonymous;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AssigneeInfoComponent.prototype, "setAnonymousText", {
        set: /**
         * @param {?} anonymousText
         * @return {?}
         */
        function (anonymousText) {
            this.anonymousText = anonymousText;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AssigneeInfoComponent.prototype, "selectedAssignees", {
        /**
         * 已选中的人（帮助不可用）
         */
        set: /**
         * 已选中的人（帮助不可用）
         * @param {?} selectedAssignees
         * @return {?}
         */
        function (selectedAssignees) {
            !this._assigneeInfo && (this.placeholder = null);
            this._selectedAssignees = selectedAssignees;
            selectedAssignees.length && (this.defaultParticipant = this.anonymous ? this.anonymousText : selectedAssignees.map((/**
             * @param {?} user
             * @return {?}
             */
            function (user) { return user.name; })).join());
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 选择参与者
     */
    /**
     * 选择参与者
     * @return {?}
     */
    AssigneeInfoComponent.prototype.clickAssignUser = /**
     * 选择参与者
     * @return {?}
     */
    function () {
        if (this.isUseNewSelectUserHelp) {
            this.extendUserHelp = true;
            this.extendUserHelpUrl = '/platform/runtime/wf/web/wf-user-help/wf-user-help-farris/wf-user-help-farris.js' + '?v=' + new Date().getTime().toString();
        }
        /** @type {?} */
        var context = {
            singleSelect: this.singleSelect,
            allowGlobalAssign: this.allowGlobalAssign,
            assigneeUsers: this._assigneeInfo && this._assigneeInfo.assigneeUsers && this._assigneeInfo.assigneeUsers.length ? this._assigneeInfo.assigneeUsers : [],
            filter: this.filter,
            selectedUsers: this.assignees && this.assignees.length ? this.assignees : this._selectedAssignees,
            wfContext: this.wfContext
        };
        if (this.extendUserHelp) {
            if (this.openFromIframe) {
                this.showFarrisUserHelpFromWindow(context);
            }
            else {
                this.showExtendUserHelp(context);
            }
        }
        else {
            this.showExtendUserHelp(context);
        }
    };
    /**
     * 加载扩展帮助
     */
    /**
     * 加载扩展帮助
     * @private
     * @param {?} context
     * @return {?}
     */
    AssigneeInfoComponent.prototype.showExtendUserHelp = /**
     * 加载扩展帮助
     * @private
     * @param {?} context
     * @return {?}
     */
    function (context) {
        var _this = this;
        System.import(this.extendUserHelpUrl).then((/**
         * @param {?} moduleFactories
         * @return {?}
         */
        function (moduleFactories) {
            /** @type {?} */
            var moduleFactory = Object.values(moduleFactories)[0];
            /** @type {?} */
            var moduleRef = moduleFactory.create(_this.injector);
            moduleRef.instance.showHelp(context).subscribe((/**
             * @param {?} re
             * @return {?}
             */
            function (re) {
                /** @type {?} */
                var userList = re.userList;
                if (userList && userList.length) {
                    _this.assignees = userList.map((/**
                     * @param {?} user
                     * @return {?}
                     */
                    function (user) { return { id: user.id, name: user.name }; }));
                    _this.assignUserNgModule = _this.assignees.map((/**
                     * @param {?} user
                     * @return {?}
                     */
                    function (user) { return user.name; })).join();
                }
                else {
                    _this.assignees = [];
                    _this.assignUserNgModule = '';
                }
                _this.afterConfirm.emit(_this.assignees);
                _this.cd.detectChanges();
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) {
                console.error(error);
            }));
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            console.error(error);
        }));
    };
    //从window上弹出用户帮助
    //从window上弹出用户帮助
    /**
     * @private
     * @param {?} context
     * @return {?}
     */
    AssigneeInfoComponent.prototype.showFarrisUserHelpFromWindow = 
    //从window上弹出用户帮助
    /**
     * @private
     * @param {?} context
     * @return {?}
     */
    function (context) {
        var _this = this;
        /** @type {?} */
        var param = new UserHelpParam();
        if (context.wfContext) {
            param.filterId = context.wfContext.processCategoryId;
        }
        param.multiSelect = !context.singleSelect;
        /** @type {?} */
        var showOrgTab = false;
        if (context.allowGlobalAssign || (context.filter && context.filter.orgId)) {
            showOrgTab = true;
        }
        /** @type {?} */
        var farrisUnSelectedUserIds = '';
        if (context.assigneeUsers && context.assigneeUsers.length) {
            farrisUnSelectedUserIds = context.assigneeUsers.map((/**
             * @param {?} user
             * @return {?}
             */
            function (user) { return user.id; })).join();
            param.unSelectedUserIds = farrisUnSelectedUserIds;
        }
        //有参与者 + 允许全员
        if (showOrgTab && farrisUnSelectedUserIds) {
            param.showTabIds = 'candidate,recent,organization,favorites,group';
            param.activeTabId = 'candidate';
        }
        //允许全员
        if (showOrgTab && !farrisUnSelectedUserIds) {
            param.showTabIds = 'recent,organization,favorites,group';
            param.activeTabId = 'recent';
        }
        //仅有参与者
        if (!showOrgTab && farrisUnSelectedUserIds) {
            param.showTabIds = 'candidate';
            param.activeTabId = 'candidate';
        }
        if (context.filter) {
            /** @type {?} */
            var filter = context.filter;
            param.includeCurrentUser = !filter.notGetCurrentUser;
            param.orgId = filter.orgId ? filter.orgId : '*';
            param.enableChildHierarchy = filter.withAllOrg;
        }
        if (context.selectedUsers && context.selectedUsers.length) {
            param.userIds = context.selectedUsers.map((/**
             * @param {?} user
             * @return {?}
             */
            function (user) { return user.id; })).join();
        }
        /** @type {?} */
        var token = 'gspWfTask' + this.index;
        ((/** @type {?} */ (top.window))).CommonPersonnelSelect.component.farrisDialogShowInWf(param, token);
        if (!this.pscChange) {
            this.pscChange = ((/** @type {?} */ (top.window))).FARRIS_PSC.selectionsChange.subscribe((/**
             * @param {?} re
             * @return {?}
             */
            function (re) {
                if (re.token == token) {
                    /** @type {?} */
                    var userList = re.data;
                    if (userList && userList.length) {
                        _this.assignees = userList.map((/**
                         * @param {?} user
                         * @return {?}
                         */
                        function (user) { return { id: user.userId, name: user.name }; }));
                        _this.assignUserNgModule = _this.assignees.map((/**
                         * @param {?} user
                         * @return {?}
                         */
                        function (user) { return user.name; })).join();
                    }
                    else {
                        _this.assignees = [];
                        _this.assignUserNgModule = '';
                    }
                    _this.afterConfirm.emit(_this.assignees);
                    _this.cd.detectChanges();
                }
            }));
        }
    };
    //清除
    //清除
    /**
     * @return {?}
     */
    AssigneeInfoComponent.prototype.assignUserClear = 
    //清除
    /**
     * @return {?}
     */
    function () {
        this.assignees = [];
        this.assignUserNgModule = null;
        this.afterConfirm.emit([]);
    };
    AssigneeInfoComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-assignee-info',
                    template: "<!-- \u6307\u5B9A\u53C2\u4E0E\u8005\uFF08\u6307\u6D3E\uFF09 -->\r\n<div\r\n  *ngIf=\"\r\n    ifShowBlank ||\r\n    _assigneeInfo ||\r\n    (_selectedAssignees && _selectedAssignees.length)\r\n  \"\r\n  [ngClass]=\"\r\n    ifHorizontal\r\n      ? 'wf-task-handler-help-div wf-submit-item'\r\n      : 'wf-task-handler-help-div'\r\n  \"\r\n>\r\n  <div\r\n    [ngClass]=\"\r\n      ifHorizontal\r\n        ? 'wf-task-handler-help-text wf-submit-item-label'\r\n        : 'wf-task-handler-help-text'\r\n    \"\r\n    title=\"{{ activityName }}\"\r\n  >\r\n    {{ activityName ? activityName : (\"task.assignText\" | translate) }}\r\n  </div>\r\n  <div\r\n    style=\"flex: 1\"\r\n    title=\"{{ assignUserNgModule ? assignUserNgModule : placeholder }}\"\r\n  >\r\n    <input-group\r\n      [(ngModel)]=\"defaultParticipant\"\r\n      [editable]=\"false\"\r\n      [readonly]=\"disabled\"\r\n      [customCls]=\"'custom-cls'\"\r\n      [groupText]=\"groupIcon\"\r\n      [placeholder]=\"placeholder\"\r\n      [(value)]=\"assignUserNgModule\"\r\n      (clickHandle)=\"clickAssignUser()\"\r\n      (clear)=\"assignUserClear()\"\r\n    ></input-group>\r\n  </div>\r\n</div>\r\n",
                    styles: [".wf-task-handler-help-div{font-size:13px;color:#2d2f33;margin-bottom:14px}.wf-task-handler-help-div-submit{margin-bottom:14px}.wf-task-handler-help-text{padding-bottom:8px;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.wf-task-batch-selected{padding-left:8px}.wf-task-batch-rows{font-size:13px;font-weight:500;padding:0 4px;color:#2a87ff}.wf-task-batch-time{font-size:13px;font-weight:500;padding:0 4px;color:#424347}.wf-task-batch-message{display:flex;background:#edf9ff;border:1px solid rgba(196,235,255,1);border-radius:8px;margin-bottom:6px;align-items:center;font-size:12px;padding:4px 10px}.wf-submit-section{padding:4px 24px 0}.wf-submit-item{display:flex;align-items:baseline}.wf-submit-item-label{padding-right:10px;width:88px;text-align:right}.wf-task-splitter{display:flex;align-items:center;color:rgba(0,0,0,.4);font-size:12px;margin-bottom:10px}.wf-task-splitter-line{border-bottom:1px dashed rgba(0,0,0,.1);flex-grow:1}"]
                }] }
    ];
    /** @nocollapse */
    AssigneeInfoComponent.ctorParameters = function () { return [
        { type: I18nService },
        { type: NotifyService },
        { type: ChangeDetectorRef },
        { type: TaskService },
        { type: Injector }
    ]; };
    AssigneeInfoComponent.propDecorators = {
        disabled: [{ type: Input }],
        placeholder: [{ type: Input }],
        singleSelect: [{ type: Input }],
        filter: [{ type: Input }],
        ifShowBlank: [{ type: Input }],
        openFromIframe: [{ type: Input }],
        activityName: [{ type: Input }],
        allowGlobalAssign: [{ type: Input }],
        wfContext: [{ type: Input }],
        index: [{ type: Input }],
        isUseNewSelectUserHelp: [{ type: Input }],
        ifHorizontal: [{ type: Input }],
        afterConfirm: [{ type: Output }],
        assigneeInfo: [{ type: Input }],
        setAnonymous: [{ type: Input }],
        setAnonymousText: [{ type: Input }],
        selectedAssignees: [{ type: Input }]
    };
    return AssigneeInfoComponent;
}());
export { AssigneeInfoComponent };
if (false) {
    /** @type {?} */
    AssigneeInfoComponent.prototype.assignUserNgModule;
    /** @type {?} */
    AssigneeInfoComponent.prototype._assigneeInfo;
    /** @type {?} */
    AssigneeInfoComponent.prototype._selectedAssignees;
    /** @type {?} */
    AssigneeInfoComponent.prototype.assignees;
    /** @type {?} */
    AssigneeInfoComponent.prototype.anonymous;
    /** @type {?} */
    AssigneeInfoComponent.prototype.anonymousText;
    /** @type {?} */
    AssigneeInfoComponent.prototype.defaultParticipant;
    /** @type {?} */
    AssigneeInfoComponent.prototype.pscChange;
    /** @type {?} */
    AssigneeInfoComponent.prototype.disabled;
    /** @type {?} */
    AssigneeInfoComponent.prototype.placeholder;
    /** @type {?} */
    AssigneeInfoComponent.prototype.singleSelect;
    /** @type {?} */
    AssigneeInfoComponent.prototype.filter;
    /** @type {?} */
    AssigneeInfoComponent.prototype.ifShowBlank;
    /** @type {?} */
    AssigneeInfoComponent.prototype.openFromIframe;
    /** @type {?} */
    AssigneeInfoComponent.prototype.activityName;
    /** @type {?} */
    AssigneeInfoComponent.prototype.allowGlobalAssign;
    /** @type {?} */
    AssigneeInfoComponent.prototype.wfContext;
    /** @type {?} */
    AssigneeInfoComponent.prototype.index;
    /** @type {?} */
    AssigneeInfoComponent.prototype.isUseNewSelectUserHelp;
    /** @type {?} */
    AssigneeInfoComponent.prototype.ifHorizontal;
    /** @type {?} */
    AssigneeInfoComponent.prototype.afterConfirm;
    /** @type {?} */
    AssigneeInfoComponent.prototype.groupIcon;
    /** @type {?} */
    AssigneeInfoComponent.prototype.extendUserHelp;
    /** @type {?} */
    AssigneeInfoComponent.prototype.extendUserHelpUrl;
    /**
     * @type {?}
     * @private
     */
    AssigneeInfoComponent.prototype.i18nService;
    /**
     * @type {?}
     * @private
     */
    AssigneeInfoComponent.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    AssigneeInfoComponent.prototype.cd;
    /**
     * @type {?}
     * @private
     */
    AssigneeInfoComponent.prototype.taskService;
    /**
     * @type {?}
     * @private
     */
    AssigneeInfoComponent.prototype.injector;
}
