// @farris/kendo-binding/esm5/lib/format/datatable-kendogrid-format.js
import { Injectable } from '@angular/core';
import { DateTimeHelperService } from '@farris/ui-common/date';
import { NumberHelperService } from '@farris/ui-common/number';
import { DateUtil } from '@farris/devkit';
var FarrisKendoGridFormatService = /** @class */ (function () {
    function FarrisKendoGridFormatService(dateService, numberService) {
        this.dateService = dateService;
        this.numberService = numberService;
    }
    /**
     * 列格式化 方法
     * @param dataItem  行数据
     * @param column 列对象
     */
    FarrisKendoGridFormatService.prototype.getColumnFormat = function (dataItem, column) {
        var value = this.getValue(dataItem, column['dataField']);
        // 没有设置行编辑器
        var type = column['dataType'];
        if (type === 'enum') {
            return this.getEnumText(value, column);
        }
        else if (type === 'boolean') {
            return value ? '是' : '否';
        }
        else if (type === 'date') {
            return this.getFormatDate(value, column);
        }
        else if (type === 'number') {
            return this.getFormatNumber(value, column);
        }
        else {
            // 格式化多行文本  和string
            return this.getMultiText(value, column);
        }
    };
    /**
     * 根据id获取value
     * @param dataItem  行数据
     * @param field id,也可能是name.Name这种形式(关联对象id或者udt)
     */
    FarrisKendoGridFormatService.prototype.getValue = function (dataItem, field) {
        var index = field.indexOf('.');
        if (index < 0) {
            return dataItem[field];
        }
        var fields = field.split('.');
        // field字符串组合
        var a = fields.reduce(function (result, next) {
            return result[next];
        }, dataItem);
        return a;
    };
    /**
     * 枚举值格式化
     */
    FarrisKendoGridFormatService.prototype.getEnumText = function (value, column) {
        var enumData = column.enumData;
        if (!enumData) {
            return;
        }
        var tempColumnText;
        enumData.forEach(function (ele) {
            if (ele.value === value) {
                tempColumnText = ele.name;
            }
        });
        return tempColumnText;
    };
    /**
     * 日期格式化
     * @param value  单元格值
     * @param column 列对象
     */
    FarrisKendoGridFormatService.prototype.getFormatDate = function (value, column) {
        if (DateUtil.isEmptyDateString(value) === true) {
            return '';
        }
        var dateFormat = column.format;
        if (!dateFormat) {
            return value;
        }
        return this.dateService.formatTo(value, dateFormat);
    };
    /**
     * 数字格式化
     * @param value  单元格值
     * @param column 列对象
     * @todo 数字格式化还不完善  这个只是金额形式的格式化;   初始化value===null时,应该显示为'' 这个还没有
     */
    FarrisKendoGridFormatService.prototype.getFormatNumber = function (value, column) {
        var precision;
        if (column && column.format) {
            precision = parseInt(column.format.substr(1), 10);
        }
        else if (column && column.editor) {
            precision = column.editor.precision;
        }
        if (isNaN(precision)) {
            precision = 2;
        }
        if (!value && value !== 0) {
            return '';
        }
        var opts = {
            prefix: column.prefix ? column.prefix : '',
            suffix: column.suffix ? column.suffix : '',
            precision: precision,
            thousand: ',',
            stripZeros: false,
        };
        return this.numberService.formatMoney(value, opts);
    };
    /**
     * 多行文本格式化 shift+enter换行  失去焦点保留原格式
     * @param value  单元格值
     * @param column 列对象
     */
    FarrisKendoGridFormatService.prototype.getMultiText = function (value, column) {
        var strValue = value;
        if (strValue !== '' && !strValue) {
            strValue = '';
        }
        if (typeof strValue !== 'string') {
            var valueType = typeof strValue;
            throw new Error(valueType + "\u7C7B\u578B\u503C\u4E0D\u5408\u6CD5\uFF0C\u65E0\u6CD5\u8FDB\u884C\u683C\u5F0F\u5316");
        }
        // 防止xss
        strValue = strValue.replace(/&/g, '&amp;')
            .replace(/</g, '&lt;')
            .replace(/>/g, '&gt;')
            .replace(/\"/g, '&quot;')
            .replace(/\'/g, '&#39;')
            .replace(/\//g, '&#x2F;');
        // 如果是多行文本框  保留格式
        if (column.editor && column.editor.type === 'MultiTextBox') {
            var INDEX = strValue.indexOf('\n');
            if (INDEX > -1) {
                return strValue.replace(/\n/g, '<br>');
            }
            return strValue;
        }
        // 若为string
        return strValue;
    };
    FarrisKendoGridFormatService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FarrisKendoGridFormatService.ctorParameters = function () { return [
        { type: DateTimeHelperService },
        { type: NumberHelperService }
    ]; };
    return FarrisKendoGridFormatService;
}());
export { FarrisKendoGridFormatService };
