// @farris/kendo-binding/esm5/lib/format/dateformat-placeholder.pipe.js
import { Pipe } from '@angular/core';
import { ViewModel, DateUtil } from '@farris/devkit';
var DateFormatPlaceholderPipe = /** @class */ (function () {
    function DateFormatPlaceholderPipe(viewModel) {
        this.viewModel = viewModel;
    }
    DateFormatPlaceholderPipe.prototype.transform = function (format, controlName) {
        if (!this.viewModel.form || !this.viewModel.form.get(controlName)) {
            return {
                year: '年',
                month: '月',
                day: '日'
            };
        }
        var value = this.viewModel.form.get(controlName).value;
        var dateStr = value;
        if (dateStr) {
            dateStr = value.toISOString();
        }
        if ((DateUtil.isEmptyDateString(dateStr) === true) && format === 'yyyy年MM月dd日') {
            return {
                year: ' ',
                month: ' ',
                day: ' '
            };
        }
        return {
            year: '年',
            month: '月',
            day: '日'
        };
    };
    DateFormatPlaceholderPipe.decorators = [
        { type: Pipe, args: [{ name: 'dateFormatPlaceholder' },] }
    ];
    /** @nocollapse */
    DateFormatPlaceholderPipe.ctorParameters = function () { return [
        { type: ViewModel }
    ]; };
    return DateFormatPlaceholderPipe;
}());
export { DateFormatPlaceholderPipe };
