// @farris/kendo-binding/esm5/lib/format/kendogrid-filter.js
import * as tslib_1 from "tslib";
import { Directive, Input, HostListener } from '@angular/core';
import { DropDownListComponent } from '@progress/kendo-angular-dropdowns';
import { FilterService, BaseFilterCellComponent } from '@progress/kendo-angular-grid';
var KendoGridFilterDirective = /** @class */ (function (_super) {
    tslib_1.__extends(KendoGridFilterDirective, _super);
    function KendoGridFilterDirective(dropdownList, filterService) {
        var _this = _super.call(this, filterService) || this;
        _this.dropdownList = dropdownList;
        _this.filterService = filterService;
        return _this;
    }
    Object.defineProperty(KendoGridFilterDirective.prototype, "selectedValue", {
        get: function () {
            var filter = this.filterByField(this.column.dataField);
            return filter ? filter.value : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(KendoGridFilterDirective.prototype, "defaultItem", {
        get: function () {
            return {
                value: null,
                name: '全部'
            };
        },
        enumerable: true,
        configurable: true
    });
    KendoGridFilterDirective.prototype.ngOnInit = function () {
        this.dropdownList.valueField = 'value';
        this.dropdownList.textField = 'name';
        this.dropdownList.valuePrimitive = true;
        this.dropdownList.defaultItem = this.defaultItem;
        this.dropdownList.value = this.selectedValue;
        if (this.column.dataType === 'enum') {
            this.dropdownList.data = this.column.enumData;
        }
        else if (this.column.dataType === 'boolean') {
            this.dropdownList.data = [{ value: true, name: '是' }, { value: false, name: '否' }];
        }
    };
    KendoGridFilterDirective.prototype.onChange = function (value) {
        this.applyFilter(value === null ?
            this.removeFilter(this.column.dataField) :
            this.updateFilter({
                field: this.column.dataField,
                operator: 'eq',
                value: value
            }));
    };
    /*
   * 获取 单元格值  主从表支持
   * @param dataItem 行数据
   * @param field 字段
   */
    KendoGridFilterDirective.prototype.getValue = function (dataItem, field) {
        var index = field.indexOf('.');
        if (index < 0) {
            return dataItem[field];
        }
        var fields = field.split('.');
        // field字符串组合
        return fields.reduce(function (result, next) {
            return result[next];
        }, dataItem);
    };
    KendoGridFilterDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[kendogridFilter]'
                },] }
    ];
    /** @nocollapse */
    KendoGridFilterDirective.ctorParameters = function () { return [
        { type: DropDownListComponent },
        { type: FilterService }
    ]; };
    KendoGridFilterDirective.propDecorators = {
        filter: [{ type: Input }],
        column: [{ type: Input }],
        onChange: [{ type: HostListener, args: ['valueChange', ['$event'],] }]
    };
    return KendoGridFilterDirective;
}(BaseFilterCellComponent));
export { KendoGridFilterDirective };
