// @gsp-cmp/ubcc-param-data-ui/esm2015/lib/service/http-service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/http-service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { SessionService } from '@ecp-caf/caf-common';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "@ecp-caf/caf-common";
/** @type {?} */
const UBCC_PARAM_TEMPLATE_ROUTER = '/api/runtime/bcc/v1.0/param/template/definition';
/** @type {?} */
const UBCC_PARAM_TEMPLATE_GET_BY_ID = `${UBCC_PARAM_TEMPLATE_ROUTER}/{id}`;
/** @type {?} */
const UBCC_PARAM_DATA_ROUTER = '/api/runtime/bcc/v1.0/param/data';
/** @type {?} */
const UBCC_PARAM_DATA = UBCC_PARAM_DATA_ROUTER;
/** @type {?} */
const UBCC_PARAM_DATA_DIMENSION = `${UBCC_PARAM_DATA_ROUTER}/dimensionalData`;
/** @type {?} */
const UBCC_PARAM_DATA_GET_BY_ID = `${UBCC_PARAM_DATA_ROUTER}/{templateId}/{bizTypeId}`;
/** @type {?} */
const UBCC_DIMENSION_PARAM_DATA_GET_BY_ID = `${UBCC_PARAM_DATA_ROUTER}/{templateId}/{bizTypeId}/dimensions/paramdatas`;
/** @type {?} */
const UBCC_PARAM_HELP_DATA_ROUTER = `/api/runtime/bcc/v1.0/param/help/{helpId}/{dataIds}`;
export class UbccParamDataManangerService {
    /**
     * @param {?} http
     * @param {?} cache
     */
    constructor(http, cache) {
        this.http = http;
        this.cache = cache;
        this.httpHeaders = new HttpHeaders();
        this.httpHeaders.set("sessionid", this.cache.getUserSessionId());
    }
    /**
     * @param {?} id
     * @return {?}
     */
    getParamTemplate(id) {
        /** @type {?} */
        let url = UBCC_PARAM_TEMPLATE_GET_BY_ID.replace('{id}', id);
        return this.http.get(url, { headers: this.httpHeaders });
    }
    /**
     * @param {?} templateId
     * @param {?} bizTypeId
     * @return {?}
     */
    getParamData(templateId, bizTypeId) {
        /** @type {?} */
        let url = UBCC_PARAM_DATA_GET_BY_ID.replace('{templateId}', templateId).replace('{bizTypeId}', bizTypeId);
        return this.http.get(url, { headers: this.httpHeaders });
    }
    /**
     * @param {?} templateId
     * @param {?} bizTypeId
     * @param {?} dimension1
     * @param {?} dimension2
     * @param {?} dimensionCount
     * @return {?}
     */
    getDimensionData(templateId, bizTypeId, dimension1, dimension2, dimensionCount) {
        /** @type {?} */
        let url = UBCC_DIMENSION_PARAM_DATA_GET_BY_ID.replace('{templateId}', templateId).replace('{bizTypeId}', bizTypeId);
        if (dimensionCount === 1) {
            url = `${url}?dimension1=${dimension1}&dimensionCount=${dimensionCount}`;
        }
        if (dimensionCount === 2) {
            url = `${url}?dimension1=${dimension1}&dimension2=${dimension2}&dimensionCount=${dimensionCount}`;
        }
        return this.http.get(url, { headers: this.httpHeaders });
    }
    /**
     * @param {?} templateId
     * @param {?} bizTypeId
     * @param {?} dimensionValues
     * @param {?} paramValues
     * @return {?}
     */
    saveDimensionAndParamData(templateId, bizTypeId, dimensionValues, paramValues) {
        /** @type {?} */
        let url = UBCC_PARAM_DATA_DIMENSION;
        /** @type {?} */
        let body = {
            templateId: templateId,
            bizTypeId: bizTypeId,
            dimensionValues: dimensionValues,
            paramValues: paramValues
        };
        return this.http.post(url, body, { headers: this.httpHeaders });
    }
    /**
     * @param {?} templateId
     * @param {?} bizTypeId
     * @param {?} paramValues
     * @return {?}
     */
    saveParamData(templateId, bizTypeId, paramValues) {
        /** @type {?} */
        let url = UBCC_PARAM_DATA;
        /** @type {?} */
        let body = {
            templateId: templateId,
            bizTypeId: bizTypeId,
            paramValues: paramValues
        };
        return this.http.post(url, body, { headers: this.httpHeaders });
    }
    /**
     * @param {?} helpId
     * @param {?} dataIds
     * @return {?}
     */
    getHelpData(helpId, dataIds) {
        /** @type {?} */
        let url = UBCC_PARAM_HELP_DATA_ROUTER.replace('{helpId}', helpId).replace('{dataIds}', dataIds.join(','));
        return this.http.get(url, { headers: this.httpHeaders });
    }
}
UbccParamDataManangerService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
UbccParamDataManangerService.ctorParameters = () => [
    { type: HttpClient },
    { type: SessionService }
];
/** @nocollapse */ UbccParamDataManangerService.ngInjectableDef = i0.defineInjectable({ factory: function UbccParamDataManangerService_Factory() { return new UbccParamDataManangerService(i0.inject(i1.HttpClient), i0.inject(i2.SessionService)); }, token: UbccParamDataManangerService, providedIn: "root" });
if (false) {
    /** @type {?} */
    UbccParamDataManangerService.prototype.httpHeaders;
    /**
     * @type {?}
     * @private
     */
    UbccParamDataManangerService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    UbccParamDataManangerService.prototype.cache;
}
