// @gsp-lcm/metadata-selector/esm5/metadata/metadata-select/components/metadatapackagelist/metadatapackagelist.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/metadatapackagelist/metadatapackagelist.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { MetadataService } from '../../metadata.service';
import { process } from '@progress/kendo-data-query';
var MetadatapackagelistComponent = /** @class */ (function () {
    function MetadatapackagelistComponent(metadataService) {
        this.metadataService = metadataService;
        // tslint:disable-next-line:member-ordering
        this.state = {
            skip: 0,
            take: 9999,
            // Initial filter descriptor
            filter: {
                logic: 'and',
                filters: []
            }
        };
        // tslint:disable-next-line:member-ordering
        this.selectableSetting = { mode: 'single', checkboxOnly: false };
    }
    /**
     * @param {?} state
     * @return {?}
     */
    MetadatapackagelistComponent.prototype.dataStateChange = /**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this.state = state;
        this.gridData = process(this.data, this.state);
    };
    /**
     * @return {?}
     */
    MetadatapackagelistComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.metadataService.GetMetadataPackages().subscribe((/**
         * @param {?} packages
         * @return {?}
         */
        function (packages) {
            /** @type {?} */
            var tempdata = [];
            _this.packagesInfo = packages;
            if (_this.packagesInfo) {
                _this.packagesInfo.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    if (item.type === 'folder' || (item.type === 'file' && item.name.endsWith('mdpkg'))) {
                        tempdata.push({ 'name': item.name });
                    }
                }));
                _this.data = tempdata;
                _this.gridData = process(_this.data, _this.state);
            }
        }));
    };
    /**
     * @param {?} event
     * @return {?}
     */
    MetadatapackagelistComponent.prototype.onSelectionChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var packageName = event.selectedRows[0].dataItem.name;
        this.metadataService.clearPackage();
        this.metadataService.addPackage(packageName);
    };
    MetadatapackagelistComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'app-metadatapackagelist',
                    template: "<div class=\"templateInfo\">\r\n  <div class=\"templateSelector\">\r\n    <kendo-grid [data]=\"gridData\" \r\n                [selectable]=\"selectableSetting\" \r\n                (selectionChange)=onSelectionChange($event)\r\n                [pageSize]=\"state.take\"\r\n                [skip]=\"state.skip\"\r\n                [sort]=\"state.sort\"\r\n                [filter]=\"state.filter\"\r\n                [sortable]=\"true\"\r\n                [pageable]=\"false\"\r\n                [filterable]=\"true\"\r\n                (dataStateChange)=dataStateChange($event)>\r\n      <kendo-grid-column field=\"name\" title=\"\u5305\u540D\u79F0\" width=\"200\">\r\n      </kendo-grid-column>\r\n    </kendo-grid>\r\n  </div>\r\n</div>",
                    styles: [".templateSelector{width:580px;height:450px;border:.5px solid #ddd}.templateDescription{width:200px;height:450px;border:.5px solid #ddd}.templateInfo{display:flex}"]
                }] }
    ];
    /** @nocollapse */
    MetadatapackagelistComponent.ctorParameters = function () { return [
        { type: MetadataService }
    ]; };
    return MetadatapackagelistComponent;
}());
export { MetadatapackagelistComponent };
if (false) {
    /** @type {?} */
    MetadatapackagelistComponent.prototype.data;
    /** @type {?} */
    MetadatapackagelistComponent.prototype.packagesInfo;
    /** @type {?} */
    MetadatapackagelistComponent.prototype.state;
    /** @type {?} */
    MetadatapackagelistComponent.prototype.gridData;
    /** @type {?} */
    MetadatapackagelistComponent.prototype.selectableSetting;
    /**
     * @type {?}
     * @private
     */
    MetadatapackagelistComponent.prototype.metadataService;
}
