// @gsp-wf/wf-process-editor/esm2015/lib/views/assign-performer-config/assign-performer-config.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, Input, ViewChild, EventEmitter, TemplateRef, HostBinding, Optional, ChangeDetectorRef, Inject, LOCALE_ID } from '@angular/core';
import { FormGroup, FormControl } from '@angular/forms';
import { AssignPerformerConfig } from './assign-performer-config.entity';
import { OrganizationPotentialOwner, ActionScope } from '../organization-potential-owner/actionOrganizationScope';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../i18n/pipe/wf-process-editor-locale.pipe';
export class AssignPerformerConfigComponent {
    /**
     * @param {?} cdr
     * @param {?} localeId
     * @param {?} resources
     * @param {?} localePipe
     */
    constructor(cdr, localeId, resources, localePipe) {
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.cls = 'farris-main-area flex-column';
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.allowGlobalAssign = false;
        this.operatorCollection = [
            { text: this.localePipe.transform('component.assignPerformer.currentUser'), value: 'currentUser' },
            { text: this.localePipe.transform('component.assignPerformer.procInstCreator'), value: 'procInstCreator' }
        ];
        this.operator = 'currentUser';
        this.organizationLevelCollection = [
            { value: 'GLOBAL', text: this.localePipe.transform('component.assignPerformer.global'), orgId: 'global' },
            { value: 'UNIT', text: this.localePipe.transform('component.assignPerformer.currentUnit'), orgId: 'currentUnit' },
            { value: 'DEPARTMENT', text: this.localePipe.transform('component.assignPerformer.currentDepartment'), orgId: 'currentDepartment' }
        ];
        this.organizationLevel = this.organizationLevelCollection[0].value;
        this.includeChildren = false;
        this.modalConfig = {
            title: this.localePipe.transform('component.assignPerformer.assignPerformer'),
            width: 720,
            height: 490,
            showButtons: true
        };
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
    }
    /**
     * @param {?} v
     * @return {?}
     */
    set value(v) {
        this.assingPerformerConfigForm = new FormGroup({
            multiSelect: new FormControl(v.multiSelect),
            necessary: new FormControl(v.necessary)
        });
        this.operator = v.operator || 'currentUser';
        this.allowGlobalAssign = v.allowGlobalAssign || false;
        if (v.assignScope) {
            this.organizationLevel = v.assignScope.organizationLevel;
            this.includeChildren = v.assignScope.includeChildren;
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    clickOK() {
        this.config = {
            operator: this.operator,
            multiSelect: this.assingPerformerConfigForm.get('multiSelect').value,
            necessary: this.assingPerformerConfigForm.get('necessary').value,
            allowGlobalAssign: this.allowGlobalAssign
        };
        if (this.allowGlobalAssign && this.organizationLevel !== this.organizationLevelCollection[0].value) {
            /** @type {?} */
            const potential = new OrganizationPotentialOwner();
            potential.actionScope = ActionScope.AssignUser;
            potential.organizationLevel = (/** @type {?} */ (this.organizationLevel));
            potential.organizationId = this.organizationLevelCollection.find((/**
             * @param {?} o
             * @return {?}
             */
            o => o.value === this.organizationLevel)).orgId;
            potential.organizationName = this.organizationLevelCollection.find((/**
             * @param {?} o
             * @return {?}
             */
            o => o.value === this.organizationLevel)).text;
            potential.includeChildren = this.includeChildren;
            potential.tag = 'variable';
            this.config.assignScope = potential;
        }
        /** @type {?} */
        const object = { value: this.config, parameters: null };
        this.submitModal.emit(object);
    }
    /**
     * @return {?}
     */
    clickCancel() {
        this.closeModal.emit();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    valueChange(event) {
        switch (event.value) {
            case this.organizationLevelCollection[0].value:
                this.includeChildren = false;
                break;
            case this.organizationLevelCollection[1].value:
                this.includeChildren = true;
                break;
            case this.organizationLevelCollection[2].value:
                this.includeChildren = false;
                break;
        }
    }
}
AssignPerformerConfigComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-assign-performer-config',
                template: "<div class=\"f-page\">\r\n  <section class=\"f-page-main d-flex flex-column\">\r\n    <farris-section class=\"pt-0 f-section-form f-section-in-mainsubcard\" mainTitle=\"{{'component.assignPerformer.assignPerformer'|wfLocale}}\" [enableAccordion]=\"'default'\"\r\n      [enableMaximize]=\"false\">\r\n      <div class=\"f-form-layout farris-form farris-form-controls-inline\">\r\n        <fieldset class=\"f-section-formgroup\">\r\n          <div class=\"f-section-formgroup-inputs\">\r\n            <div class=\"col-12\">\r\n              <div class=\"farris-group-wrap\">\r\n                <div class=\"form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-text\">{{'component.assignPerformer.operator'|wfLocale}}</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap\" style=\"width: 300px;\">\r\n                    <farris-combo-list [(ngModel)]=\"operator\" name=\"operator\" [ngModelOptions]=\"{standalone: true}\"\r\n                      [idField]=\"'value'\" [textField]=\"'text'\" [valueField]=\"'value'\" [data]=\"operatorCollection\"\r\n                      [panelHeight]=\"60\">\r\n                    </farris-combo-list>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n            <div class=\"col-6\">\r\n              <div class=\"farris-group-wrap ml-2 pl-4\">\r\n                <div class=\"form-group farris-form-group\">\r\n                  <div class=\"farris-input-wrap\">\r\n                    <div class=\"custom-control custom-checkbox\">\r\n                      <input class=\"custom-control-input\" type=\"checkbox\" id=\"multiSelect\" formControlName=\"multiSelect\"\r\n                        [readOnly]=\"false\" />\r\n                      <label class=\"custom-control-label\" for=\"multiSelect\">{{'component.assignPerformer.multiSelect'|wfLocale}}</label>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n            <div class=\"col-6\">\r\n              <div class=\"farris-group-wrap\">\r\n                <div class=\"form-group farris-form-group\">\r\n                  <div class=\"farris-input-wrap\">\r\n                    <div class=\"custom-control custom-checkbox\">\r\n                      <input class=\"custom-control-input\" type=\"checkbox\" id=\"necessary\" formControlName=\"necessary\" [readOnly]=\"false\" />\r\n                      <label class=\"custom-control-label\" for=\"necessary\">{{'component.assignPerformer.necessary'|wfLocale}}</label>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </fieldset>\r\n      </div>\r\n    </farris-section>\r\n    <!-- <farris-section class=\"pt-0 f-section-form f-section-in-mainsubcard\" [mainTitle]=\"'\u5907\u9009\u8303\u56F4'\" [enableAccordion]=\"'default'\"\r\n      [enableMaximize]=\"false\">\r\n      <div class=\"f-form-layout farris-form farris-form-controls-inline\">\r\n        <fieldset class=\"f-section-formgroup\">\r\n          <div class=\"f-section-formgroup-inputs\">\r\n            <div class=\"col-12\">\r\n              <div class=\"farris-group-wrap ml-2 pl-4\">\r\n                <div class=\"form-group\">\r\n                  <div class=\"farris-input-wrap\" style=\"width: 300px;\">\r\n                    <farris-radiogroup [(ngModel)]=\"assignScope\" [data]=\"assignScopeList\" name=\"assignScope\"\r\n                      [horizontal]=\"true\">\r\n                    </farris-radiogroup>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </fieldset>\r\n      </div>\r\n    </farris-section>\r\n    <farris-section class=\"pt-0 f-section-grid f-section-in-managelist\" [mainTitle]=\"'\u5907\u9009\u8303\u56F4'\" [enableAccordion]=\"'default'\"\r\n      [enableMaximize]=\"false\" style=\"height: 200px;\">\r\n      <app-performer [type]=\"'assignPerformer'\"></app-performer>\r\n    </farris-section> -->\r\n  </section>\r\n</div>\r\n\r\n<!--\r\n<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <div class=\"f-section-header pt-2 mx-2\">\r\n        <div class=\"f-title\">\r\n          <h4 class=\"f-title-text\">\u6307\u6D3E\u89C4\u5219\u9009\u9879</h4>\r\n        </div>\r\n      </div>\r\n      <div class=\"f-utils-fill-flex-column h-100\">\r\n        <form class=\"h-100\" [formGroup]=\"assingPerformerConfigForm\">\r\n          <div class=\"f-utils-fill-flex-column justify-content-center h-100 pb-4\">\r\n            <div class=\"my-2 mx-5\">\r\n              <div class=\"farris-group-wrap\">\r\n                <div class=\"form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-text\">\u6307\u6D3E\u64CD\u4F5C\u4EBA</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap mt-1\" style=\"width: 300px;\">\r\n                    <farris-combo-list [(ngModel)]=\"operator\" name=\"operator\" [ngModelOptions]=\"{standalone: true}\"\r\n                      [idField]=\"'value'\" [textField]=\"'text'\" [valueField]=\"'value'\" [data]=\"operatorCollection\"\r\n                      [panelHeight]=\"60\">\r\n                    </farris-combo-list>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n            <div class=\"my-2 mx-5\">\r\n              <div class=\"farris-input-wrap\">\r\n                <div class=\"custom-control custom-checkbox\">\r\n                  <input class=\"custom-control-input\" type=\"checkbox\" id=\"multiSelect\" formControlName=\"multiSelect\"\r\n                    [readOnly]=\"false\" />\r\n                  <label class=\"custom-control-label\" for=\"multiSelect\">\u5141\u8BB8\u9009\u62E9\u591A\u4E2A\u5F85\u9009\u7528\u6237</label>\r\n                </div>\r\n              </div>\r\n            </div>\r\n            <div class=\"my-2 mx-5\">\r\n              <div class=\"farris-input-wrap\">\r\n                <div class=\"custom-control custom-checkbox\">\r\n                  <input class=\"custom-control-input\" type=\"checkbox\" id=\"necessary\" formControlName=\"necessary\" [readOnly]=\"false\" />\r\n                  <label class=\"custom-control-label\" for=\"necessary\">\u53EA\u6709\u4E00\u4E2A\u5F85\u9009\u7528\u6237\u65F6\uFF0C\u4ECD\u9700\u624B\u5DE5\u9009\u62E9</label>\r\n                </div>\r\n              </div>\r\n            </div>\r\n            <div class=\"mx-2 mt-3\">\r\n              <div class=\"title-content\">\r\n                <h4 class=\"title-text\">\u8BBE\u5B9A\u7684\u53C2\u4E0E\u8005\u8303\u56F4\u5916\u6307\u6D3E</h4><span class=\"title-line\"></span>\r\n              </div>\r\n            </div>\r\n            <div class=\"farris-form form-inline farris-form-inline my-1\">\r\n              <div class=\"col-12 mt-3\">\r\n                <div class=\"farris-group-wrap\">\r\n                  <div class=\"form-group farris-form-group\">\r\n                    <label class=\"col-form-label\">\r\n                      <span class=\"farris-label-text\">\u662F\u5426\u542F\u7528</span>\r\n                    </label>\r\n                    <div class=\"farris-input-wrap ml-1\">\r\n                      <farris-switch [(ngModel)]=\"allowGlobalAssign\" [ngModelOptions]=\"{standalone: true}\" name=\"allowGlobalAssign\"></farris-switch>\r\n                      <div class=\"farris-feedback valid-feedback\"></div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n              <div class=\"d-flex flex-row\">\r\n                <div class=\"col-8 my-2\">\r\n                  <div class=\"farris-group-wrap\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                      <label class=\"col-form-label\">\r\n                        <span class=\"farris-label-text\">\u7EC4\u7EC7\u8303\u56F4</span>\r\n                      </label>\r\n                      <div class=\"farris-input-wrap ml-1\">\r\n                        <farris-combo-list [panelHeight]=\"100\" [(ngModel)]=\"organizationLevel\" [ngModelOptions]=\"{standalone: true}\"\r\n                          name=\"organizationLevel\" [idField]=\"'value'\" [textField]=\"'text'\" [valueField]=\"'value'\"\r\n                          [data]=\"organizationLevelCollection\" [readonly]=\"allowGlobalAssign===false\" (valueChange)=\"valueChange($event)\">\r\n                        </farris-combo-list>\r\n                        <div class=\"farris-feedback valid-feedback\"></div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n                <div class=\"col-4 my-2\">\r\n                  <div class=\"farris-input-wrap pt-1\">\r\n                    <div class=\"custom-control custom-checkbox mr-3\">\r\n                      <input class=\"custom-control-input\" type=\"checkbox\" [(ngModel)]=\"includeChildren\" [ngModelOptions]=\"{standalone: true}\"\r\n                        name=\"includeChildren\" id=\"includeChildren\" [disabled]=\"!(allowGlobalAssign===true && organizationLevel === organizationLevelCollection[2].value)\" />\r\n                      <label class=\"custom-control-label\" for=\"includeChildren\">\u5305\u542B\u4E0B\u7EA7</label>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </form>\r\n      </div>\r\n    </section>\r\n  </div>\r\n</div> -->\r\n\r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm px-3 mx-2\" (click)=\"clickCancel()\">{{'component.assignPerformer.clickCancel'|wfLocale}}</button>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm px-3 mx-2\" (click)=\"clickOK()\">{{'component.assignPerformer.clickOK'|wfLocale}}</button>\r\n</ng-template>\r\n\r\n",
                providers: [
                    WfLocalePipe,
                ],
                styles: [".title-content{color:#000;display:inline-flex;align-items:center;line-height:1.375rem;position:relative;overflow:hidden;padding:0 0 0 1rem;min-width:415px}.title-content ::before{content:'';width:.25rem;height:1.125rem;background:#59a1ff;border-radius:1px;position:absolute;top:50%;left:0;margin-top:-.5625rem}.title-text{font-size:1rem;line-height:1.375rem;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;margin-bottom:0;margin-right:10px}.title-line{flex-grow:1;border-bottom:1px dashed rgba(0,0,0,.1)}"]
            }] }
];
/** @nocollapse */
AssignPerformerConfigComponent.ctorParameters = () => [
    { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
    { type: WfLocalePipe, decorators: [{ type: Optional }] }
];
AssignPerformerConfigComponent.propDecorators = {
    cls: [{ type: HostBinding, args: ['class',] }],
    closeModal: [{ type: Output }],
    submitModal: [{ type: Output }],
    value: [{ type: Input }],
    modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }]
};
if (false) {
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.cls;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.closeModal;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.submitModal;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.assingPerformerConfigForm;
    /**
     * @type {?}
     * @private
     */
    AssignPerformerConfigComponent.prototype.config;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.modalFooter;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.allowGlobalAssign;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.operatorCollection;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.operator;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.organizationLevelCollection;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.organizationLevel;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.includeChildren;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.modalConfig;
    /**
     * @type {?}
     * @private
     */
    AssignPerformerConfigComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    AssignPerformerConfigComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    AssignPerformerConfigComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    AssignPerformerConfigComponent.prototype.localePipe;
}
export class AssignConfigConverter {
    /**
     * @param {?} localePipe
     */
    constructor(localePipe) {
        this.localePipe = localePipe;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    convertTo(data) {
        return this.localePipe.transform('component.assignPerformer.set');
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    AssignConfigConverter.prototype.localePipe;
}
