// @farris/command-services/esm5/lib/authority/authority.js
import { empty } from 'rxjs';
import { map } from 'rxjs/operators';
/**
 * 权限
 */
var Authority = /** @class */ (function () {
    function Authority() {
        this.operationLabelMap = new Map();
        this.authOfLabel = new Map();
    }
    /**
     * 设置label到operation id的映射
     * @param labelMap
     */
    Authority.prototype.setLabelMap = function (labelMap) {
        var _this = this;
        for (var i in labelMap) {
            this.operationLabelMap.set(i, labelMap[i]);
        }
        this.operationLabelMap.forEach(function (opId, label) {
            _this[label] = false;
        });
    };
    /**
     * 初始化权限的值
     */
    Authority.prototype.initialize = function (authorityService) {
        var _this = this;
        if (this.operationLabelMap && this.operationLabelMap.size > 0) {
            return authorityService.queryOperationAuthority(Array.from(this.operationLabelMap.values())).pipe(map(function (operationAuth) {
                _this.operationLabelMap.forEach(function (opId, label) {
                    _this.authOfLabel.set(label, operationAuth.has(opId) ? operationAuth.get(opId) : false);
                });
            }));
        }
        else {
            return empty();
        }
    };
    /**
     * 返回操作权限，即用户是否拥有Label所对应的权限
     * @param label 标签，可以是button id等
     */
    Authority.prototype.hasOperation = function (label) {
        return this.authOfLabel.has(label) && this.authOfLabel.get(label);
    };
    return Authority;
}());
export { Authority };
