// @qdp/common/esm2015/lib/service/print-service/print.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { OutputType, FileType, PrintType } from '@gsp-svc/cloudprint';
/**
 * 打印服务
 * \@Scope FrameComponent
 */
export class IntegrationPrintService {
    /**
     * 构造函数
     * @param {?} _printService
     */
    constructor(_printService) {
        this.printService = _printService;
    }
    /**
     * 打印单据
     * @param {?} beMetaId BE元数据标识
     * @param {?} bizBillId 业务单据标识
     * @return {?}
     */
    printSingle(beMetaId, bizBillId) {
        return this.printArray(beMetaId, [bizBillId]);
    }
    /**
     * 打印单据（通过id或ids）
     * @param {?} beMetaId be元数据Id
     * @param {?} ids 单据id或ids
     * @return {?}
     */
    printByIds(beMetaId, ids) {
        /** @type {?} */
        const dataIds = ids.split(',').filter((/**
         * @param {?} item
         * @return {?}
         */
        item => item));
        return this.printArray(beMetaId, dataIds);
    }
    /**
     * 打印单据（带维度）
     * @param {?} beMetaId be元数据Id
     * @param {?} ids 单据id或ids
     * @param {?=} dim1 维度1值
     * @param {?=} dim2 维度2值
     * @param {?=} billCategoryId 单据类型Id
     * @return {?}
     */
    printByIdsWithDimension(beMetaId, ids, dim1, dim2, billCategoryId) {
        /** @type {?} */
        const dataIds = ids.split(',').filter((/**
         * @param {?} item
         * @return {?}
         */
        item => item));
        return this.printArray(beMetaId, dataIds, dim1, dim2, billCategoryId);
    }
    /**
     * 打印单据（带维度）
     * @param {?} beMetaId be元数据Id
     * @param {?} ids 单据id或ids
     * @param {?} dim1 维度1值
     * @param {?} dim2 维度2值
     * @param {?} queryType
     * @param {?} queryServiceId
     * @param {?} queryParam
     * @return {?}
     */
    printWithDimensionCustomDataSource(beMetaId, ids, dim1, dim2, queryType, queryServiceId, queryParam) {
        /** @type {?} */
        const dataIds = ids.split(',').filter((/**
         * @param {?} item
         * @return {?}
         */
        item => item));
        return this.printArray(beMetaId, dataIds, dim1, dim2, undefined, queryType, queryServiceId, queryParam);
    }
    /**
     * 打印多个单据
     * @param {?} beMetaId BE元数据标识
     * @param {?} dataIds 业务单据标识数组
     * @param {?=} dim1 维度1
     * @param {?=} dim2 维度2
     * @param {?=} billCategoryId 业务单据类型Id
     * @param {?=} queryType
     * @param {?=} queryServiceId
     * @param {?=} queryParam
     * @return {?}
     */
    printArray(beMetaId, dataIds, dim1, dim2, billCategoryId, queryType, queryServiceId, queryParam) {
        /** @type {?} */
        const sourceOptions = this.buildSourceOptions({
            dataIds: dataIds,
            sourceId: beMetaId
        });
        /** @type {?} */
        const outputOptions = this.buildOutputOptions();
        if (dim1) {
            sourceOptions.FirstDimensionVal = dim1;
        }
        if (dim2) {
            sourceOptions.SecondDimensionVal = dim2;
        }
        if (typeof billCategoryId !== 'undefined') {
            sourceOptions.billCategoryId = billCategoryId;
        }
        if (queryType) {
            sourceOptions.queryType = queryType;
        }
        if (queryServiceId) {
            sourceOptions.queryServiceId = queryServiceId;
        }
        if (queryParam) {
            sourceOptions.queryParam = queryParam;
        }
        return this.printService.outputBEData(sourceOptions, outputOptions, 'tab');
    }
    /**
     * 构造SourceOptions
     * @private
     * @param {?} options options
     * @return {?}
     */
    buildSourceOptions(options) {
        /** @type {?} */
        const so = {
            DataIds: options && options.dataIds || undefined,
            SourceId: options && options.sourceId || undefined,
            FirstDimensionVal: options && options.dim1 || undefined,
            SecondDimensionVal: options && options.dim2 || undefined,
            RetrieveParam: options && options.retrieveParam || undefined,
            FormatId: options && options.formatId || undefined,
            billCategoryId: options && options.billCategoryId || undefined,
            ServiceUnit: options && options.serviceUnit || undefined,
            currentPage: options && options.currentPage || undefined,
            pageRowCount: options && options.pageRowCount || undefined,
            queryType: options && options.queryType || undefined,
            queryServiceId: options && options.queryServiceId || undefined,
            bizTypeId: options && options.bizTypeId || undefined,
            queryParam: options && options.queryParam || undefined
        };
        return so;
    }
    /**
     * 构造OutputOptions
     * @private
     * @param {?=} options options
     * @return {?}
     */
    buildOutputOptions(options) {
        /** @type {?} */
        const oo = {
            OutputType: options && options.outputType || OutputType.PRINT,
            FileType: options && options.fileType || FileType.Html5,
            Path: options && options.path || undefined,
            DeviceId: options && options.deviceId || undefined,
            printJob: options && options.printJob || undefined,
            printerName: options && options.printerName || undefined,
            printSetting: options && options.printSetting || undefined,
            printType: options && options.printType || PrintType.Form
        };
        return oo;
    }
}
if (false) {
    /** @type {?} */
    IntegrationPrintService.prototype.printService;
}
