// @gsp-wf/biz-process-api/fesm2015/gsp-wf-biz-process-api.js
import { Injectable, Component, NgModule, defineInjectable } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class BizProcessApiService {
    constructor() { }
}
BizProcessApiService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
BizProcessApiService.ctorParameters = () => [];
/** @nocollapse */ BizProcessApiService.ngInjectableDef = defineInjectable({ factory: function BizProcessApiService_Factory() { return new BizProcessApiService(); }, token: BizProcessApiService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class BizProcessApiComponent {
    constructor() { }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
BizProcessApiComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-biz-process-api',
                template: `
    <p>
      biz-process-api works!
    </p>
  `
            }] }
];
/** @nocollapse */
BizProcessApiComponent.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class BizProcessApiModule {
}
BizProcessApiModule.decorators = [
    { type: NgModule, args: [{
                declarations: [BizProcessApiComponent],
                imports: [],
                exports: [BizProcessApiComponent]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { BizProcessApiService, BizProcessApiComponent, BizProcessApiModule };
