// @gsp-svc/file-viewer/esm2015/lib/ppt-viewer/ppt-viewer/ppt-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
export class PptViewerComponent {
    /**
     * @param {?} sanitizer
     */
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.viewUrl = this.sanitizer.bypassSecurityTrustResourceUrl(this.pptUrl);
    }
    /**
     * @param {?} url
     * @return {?}
     */
    pptChange(url) {
        this.viewUrl = this.sanitizer.bypassSecurityTrustResourceUrl(url);
    }
}
PptViewerComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-ppt-viewer',
                template: "<div class=\"f-page\"><iframe id=\"viewPpt\" [src]=\"viewUrl\" height=\"100%\" width=\"100%\"></iframe></div>",
                styles: [""]
            }] }
];
/** @nocollapse */
PptViewerComponent.ctorParameters = () => [
    { type: DomSanitizer }
];
PptViewerComponent.propDecorators = {
    pptUrl: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PptViewerComponent.prototype.pptUrl;
    /** @type {?} */
    PptViewerComponent.prototype.viewUrl;
    /**
     * @type {?}
     * @private
     */
    PptViewerComponent.prototype.sanitizer;
}
