// @gsp-cmp/ccmp-ui/esm5/lib/view-utils/view-utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { MessagerService } from '@farris/ui-messager';
import { NotifyService } from '@farris/ui-notify';
import * as i0 from "@angular/core";
import * as i1 from "@farris/ui-messager";
import * as i2 from "@farris/ui-notify";
var ViewUtils = /** @class */ (function () {
    function ViewUtils(messagerService, notify) {
        this.messagerService = messagerService;
        this.notify = notify;
    }
    /**
     * 获取模态框配置信息
     * @param btnRef 按钮模板引用
     * @param title 模态框标题
     * @param options 配置项，暂未使用
     */
    /**
     * 获取模态框配置信息
     * @param {?} btnRef 按钮模板引用
     * @param {?=} options 配置项，暂未使用
     * @return {?}
     */
    ViewUtils.prototype.getModalConfig = /**
     * 获取模态框配置信息
     * @param {?} btnRef 按钮模板引用
     * @param {?=} options 配置项，暂未使用
     * @return {?}
     */
    function (btnRef, options) {
        return {
            title: (options && options.title) ? options.title : '操作列表',
            width: (options && options.width) ? options.width : 450,
            height: (options && options.height) ? options.width : 500,
            resizable: (options && options.resizable) ? options.resizable : true,
            showButtons: true,
            showMaxButton: false,
            enableScroll: false,
            buttons: btnRef
        };
    };
    /**
     * @param {?} data
     * @return {?}
     */
    ViewUtils.prototype.showObseleteDialog = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.messagerService.show('warning', "\u901A\u7528\u6784\u4EF6" + data.code + "\u5DF2\u8FC7\u65F6\uFF01\n                    \u5347\u7EA7\u6B65\u9AA4\uFF1A\n                    1. \u91CD\u65B0\u5B9A\u4E49\u901A\u7528\u6784\u4EF6\u5143\u6570\u636E;\n                    2. \u5F00\u53D1\u65B0\u7248\u672C\u6784\u4EF6\u4EE3\u7801\u3002\n                    \u8BF7\u53C2\u8003GSPCloud\u5F00\u53D1\u6587\u6863->\u9AD8\u7EA7\u5F00\u53D1->\u6784\u4EF6\u5F00\u53D1->\u901A\u7528\u6784\u4EF6\u5F00\u53D1", { safeHtml: false });
    };
    /**
     * @param {?} message
     * @param {?} infoType
     * @param {?} title
     * @param {?} timeout
     * @return {?}
     */
    ViewUtils.prototype.showInfo = /**
     * @param {?} message
     * @param {?} infoType
     * @param {?} title
     * @param {?} timeout
     * @return {?}
     */
    function (message, infoType, title, timeout) {
        this.notify.config.position = 'top-center';
        if (infoType == 'success') {
            this.notify.success((/** @type {?} */ ({ title: title, msg: message, timeout: timeout })));
        }
        else if (infoType == 'info') {
            this.notify.info((/** @type {?} */ ({ title: title, msg: message, timeout: timeout })));
        }
        else if (infoType == 'warning') {
            this.notify.warning((/** @type {?} */ ({ title: title, msg: message, timeout: timeout })));
        }
        else if (infoType == 'error') {
            this.notify.error((/** @type {?} */ ({ title: title, msg: message, timeout: timeout })));
        }
    };
    ViewUtils.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ViewUtils.ctorParameters = function () { return [
        { type: MessagerService },
        { type: NotifyService }
    ]; };
    /** @nocollapse */ ViewUtils.ngInjectableDef = i0.defineInjectable({ factory: function ViewUtils_Factory() { return new ViewUtils(i0.inject(i1.MessagerService), i0.inject(i2.NotifyService)); }, token: ViewUtils, providedIn: "root" });
    return ViewUtils;
}());
export { ViewUtils };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ViewUtils.prototype.messagerService;
    /**
     * @type {?}
     * @private
     */
    ViewUtils.prototype.notify;
}
