// @farris/ui-editor/esm2015/lib/utils/ScriptLoader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { uuid } from './Utils';
/**
 * @record
 */
export function IStateObj() { }
if (false) {
    /** @type {?} */
    IStateObj.prototype.listeners;
    /** @type {?} */
    IStateObj.prototype.scriptId;
    /** @type {?} */
    IStateObj.prototype.scriptLoaded;
}
/** @type {?} */
const createState = (/**
 * @return {?}
 */
() => {
    return {
        listeners: [],
        scriptId: uuid('tiny-script'),
        scriptLoaded: false
    };
});
const ɵ0 = createState;
// WARNING: interface has both a type and a value, skipping emit
/** @type {?} */
const CreateScriptLoader = (/**
 * @return {?}
 */
() => {
    /** @type {?} */
    let state = createState();
    /** @type {?} */
    const injectScriptTag = (/**
     * @param {?} scriptId
     * @param {?} doc
     * @param {?} url
     * @param {?} callback
     * @return {?}
     */
    (scriptId, doc, url, callback) => {
        /** @type {?} */
        const scriptTag = doc.createElement('script');
        scriptTag.referrerPolicy = 'origin';
        scriptTag.type = 'application/javascript';
        scriptTag.id = scriptId;
        scriptTag.src = url;
        /** @type {?} */
        const handler = (/**
         * @return {?}
         */
        () => {
            scriptTag.removeEventListener('load', handler);
            callback();
        });
        scriptTag.addEventListener('load', handler);
        if (doc.head) {
            doc.head.appendChild(scriptTag);
        }
    });
    /** @type {?} */
    const load = (/**
     * @param {?} doc
     * @param {?} url
     * @param {?} callback
     * @return {?}
     */
    (doc, url, callback) => {
        if (state.scriptLoaded) {
            callback();
        }
        else {
            state.listeners.push(callback);
            if (!doc.getElementById(state.scriptId)) {
                injectScriptTag(state.scriptId, doc, url, (/**
                 * @return {?}
                 */
                () => {
                    state.listeners.forEach((/**
                     * @param {?} fn
                     * @return {?}
                     */
                    (fn) => fn()));
                    state.scriptLoaded = true;
                }));
            }
        }
    });
    // Only to be used by tests.
    /** @type {?} */
    const reinitialize = (/**
     * @return {?}
     */
    () => {
        state = createState();
    });
    return {
        load,
        reinitialize
    };
});
const ɵ1 = CreateScriptLoader;
/** @type {?} */
const ScriptLoader = CreateScriptLoader();
export { ScriptLoader };
export { ɵ0, ɵ1 };
