// @farris/ui-editor/esm2015/lib/utils/Utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { validEvents } from '../Events';
/** @type {?} */
const bindHandlers = (/**
 * @param {?} ctx
 * @param {?} editor
 * @return {?}
 */
(ctx, editor) => {
    validEvents.forEach((/**
     * @param {?} eventName
     * @return {?}
     */
    (eventName) => {
        /** @type {?} */
        const eventEmitter = ctx[eventName];
        editor.on(eventName.substring(2), (/**
         * @param {?} event
         * @return {?}
         */
        (event) => ctx.ngZone.run((/**
         * @return {?}
         */
        () => eventEmitter.emit({ event, editor })))));
    }));
});
const ɵ0 = bindHandlers;
/** @type {?} */
let unique = 0;
/** @type {?} */
const uuid = (/**
 * @param {?} prefix
 * @return {?}
 */
(prefix) => {
    /** @type {?} */
    const date = new Date();
    /** @type {?} */
    const time = date.getTime();
    /** @type {?} */
    const random = Math.floor(Math.random() * 1000000000);
    unique++;
    return prefix + '_' + random + unique + String(time);
});
const ɵ1 = uuid;
/** @type {?} */
const isTextarea = (/**
 * @param {?=} element
 * @return {?}
 */
(element) => {
    return typeof element !== 'undefined' && element.tagName.toLowerCase() === 'textarea';
});
const ɵ2 = isTextarea;
/** @type {?} */
const normalizePluginArray = (/**
 * @param {?=} plugins
 * @return {?}
 */
(plugins) => {
    if (typeof plugins === 'undefined' || plugins === '') {
        return [];
    }
    return Array.isArray(plugins) ? plugins : plugins.split(' ');
});
const ɵ3 = normalizePluginArray;
/** @type {?} */
const mergePlugins = (/**
 * @param {?} initPlugins
 * @param {?=} inputPlugins
 * @return {?}
 */
(initPlugins, inputPlugins) => normalizePluginArray(initPlugins).concat(normalizePluginArray(inputPlugins)));
const ɵ4 = mergePlugins;
// tslint:disable-next-line:no-empty
/** @type {?} */
const noop = (/**
 * @return {?}
 */
() => { });
const ɵ5 = noop;
/** @type {?} */
const isNullOrUndefined = (/**
 * @param {?} value
 * @return {?}
 */
(value) => value === null || value === undefined);
const ɵ6 = isNullOrUndefined;
export { bindHandlers, uuid, isTextarea, normalizePluginArray, mergePlugins, noop, isNullOrUndefined };
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6 };
