// @farris/ui-personnel-selector/esm2015/utils/loading-service.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/loading-service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
class DataLoadingService {
    /**
     * 构造函数
     * 注入\@farris/ui的LoadingService
     */
    constructor() {
        //this.loadingService = this.injector.get(LoadingService);
    }
    /**
     * 显示加载中
     * @param {?=} configOrMessage
     * @return {?}
     */
    show(configOrMessage) {
        if (this.loadingCmp) {
            this.loadingCmp.close();
            this.loadingCmp = null;
        }
        /** @type {?} */
        const loadingConfig = this.buildLoadingConfig(configOrMessage);
        this.loadingCmp = this.loadingService.show(loadingConfig);
        return this.loadingCmp;
    }
    /**
     * 构造LoadingConfig
     * @private
     * @param {?} configOrMessage 配置对象或消息字符串
     * @return {?}
     */
    buildLoadingConfig(configOrMessage) {
        /** @type {?} */
        let loadingConfig;
        if (!!configOrMessage) {
            if (typeof configOrMessage === 'object') {
                loadingConfig = configOrMessage;
            }
            else {
                loadingConfig = { 'message': configOrMessage };
            }
        }
        else {
            loadingConfig = {};
        }
        if (!loadingConfig.hasOwnProperty('delay')) {
            loadingConfig.delay = 0;
        }
        return loadingConfig;
    }
    /**
     * 隐藏加载中
     * @return {?}
     */
    hide() {
        if (!this.loadingCmp) {
            return;
        }
        this.loadingCmp.close();
        this.loadingCmp = null;
    }
}
DataLoadingService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DataLoadingService.ctorParameters = () => [];
if (false) {
    /**
     * 加载中组件实例
     * @type {?}
     * @private
     */
    DataLoadingService.prototype.loadingCmp;
    /** @type {?} */
    DataLoadingService.prototype.loadingService;
}
export { DataLoadingService };
