// @gsp-cmp/common-component/fesm2015/gsp-cmp-common-component.js
import { CommonModule } from '@angular/common';
import { Injectable, NgModule } from '@angular/core';
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { of } from 'rxjs';
import { map, mergeMap } from 'rxjs/operators';
import { DataType, DefaultDataTypeSerializerFactory, DefaultOperationSerializerFactory, DefaultParameterSerializerFactory, DefaultPropertySerializerFactory, DefaultSerializeContext, CommonStructureDTService, CollectionType, MapType, PrimitiveType } from '@ecp-caf/common-structure';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class GspComponent {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class GspComponentConstant {
}
GspComponentConstant.CMP_STRUCTURED_TYPE_REF = "CmpStructureTypeRef";

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class StructureReferType extends DataType {
    /**
     * @return {?}
     */
    getKeys() {
        throw new Error("Method not implemented.");
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getValue(key) {
        throw new Error("Method not implemented.");
    }
    constructor() {
        super();
        this.kind = GspComponentConstant.CMP_STRUCTURED_TYPE_REF;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 引用的元数据structure信息
 */
class RefStructureInfo {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class RefStructureUtil {
    /**
     * @param {?} refInfo
     * @return {?}
     */
    static convertToObject(refInfo) {
        /** @type {?} */
        let result = new Object();
        result["id"] = refInfo.id;
        result["referId"] = refInfo.referId;
        result["referType"] = refInfo.referType;
        result["referCode"] = refInfo.referCode;
        result["structuredTypeId"] = refInfo.structuredTypeId;
        result["code"] = refInfo.code;
        result["name"] = refInfo.name;
        return result;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    static initFromObject(obj) {
        /** @type {?} */
        let result = new RefStructureInfo();
        result.id = obj['id'];
        result.referId = obj['referId'];
        result.referType = obj['referType'];
        result.referCode = obj['referCode'];
        result.structuredTypeId = obj['structuredTypeId'];
        result.code = obj['code'];
        result.name = obj['name'];
        return result;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class GspComponentConvertor {
    /**
     * @param {?} obj
     * @param {?} component
     * @param {?} ctxt
     * @return {?}
     */
    convertToObject(obj, component, ctxt) {
        obj['id'] = component.id;
        obj['code'] = component.code;
        obj['name'] = component.name;
        obj['description'] = component.description;
        obj['application'] = component.application;
        obj['serviceUnit'] = component.serviceUnit;
        obj['version'] = component.version;
        //Operations
        obj['operations'] = new Array();
        if (component.operations) {
            component.operations.forEach((/**
             * @param {?} op
             * @return {?}
             */
            op => {
                /** @type {?} */
                var kind = op['kind'];
                /** @type {?} */
                var serializer = ctxt.getOperationSerializerFactory().get(kind);
                /** @type {?} */
                var opObj = serializer.serialize(op, ctxt);
                obj['operations'].push(opObj);
            }));
        }
        //StructuredTypes
        // obj['structuredTypes'] = new Array<Object>();
        // if (component.structuredTypes) {
        //     component.structuredTypes.forEach(structuredType => {
        //         var kind=structuredType.kind;
        //         var serializer=ctxt.getDataTypeSerializerFactory().get(kind);
        //         var structuredTypeObj=serializer.serialize(structuredType,ctxt);
        //         obj['structuredTypes'].push(structuredTypeObj);
        //     });
        // }
        //refStructures
        obj['refStructures'] = new Array();
        if (component.refStructures) {
            component.refStructures.forEach((/**
             * @param {?} refInfo
             * @return {?}
             */
            refInfo => {
                /** @type {?} */
                let refObj = RefStructureUtil.convertToObject(refInfo);
                obj['refStructures'].push(refObj);
            }));
        }
    }
    /**
     * @param {?} component
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    initFromObject(component, obj, ctxt) {
        component.id = obj['id'];
        component.code = obj['code'];
        component.name = obj['name'];
        component.description = obj['description'];
        component.application = obj['application'];
        component.serviceUnit = obj['serviceUnit'];
        component.version = obj['version'];
        //operations
        if (obj['operations']) {
            component.operations = new Array();
            obj['operations'].forEach((/**
             * @param {?} operationObj
             * @return {?}
             */
            operationObj => {
                /** @type {?} */
                var kind = operationObj['kind'];
                /** @type {?} */
                var serializer = ctxt.getOperationSerializerFactory().get(kind);
                /** @type {?} */
                var operation = serializer.deserialize(operationObj, ctxt);
                component.operations.push(operation);
            }));
        }
        //structuredTypes
        // if (obj['structuredTypes']) {
        //     component.structuredTypes = new Array<StructuredType>();
        //     obj['structuredTypes'].forEach(structureObj => {
        //         var kind=structureObj['kind'];
        //         var serializer=ctxt.getDataTypeSerializerFactory().get(kind);
        //         var structuredType=serializer.deserialize(structureObj,ctxt) as StructuredType;
        //         component.structuredTypes.push(structuredType);
        //     });
        // }
        //refStructures
        if (obj['refStructures']) {
            component.refStructures = new Array();
            obj['refStructures'].forEach((/**
             * @param {?} refObj
             * @return {?}
             */
            refObj => {
                /** @type {?} */
                let refInfo = RefStructureUtil.initFromObject(refObj);
                component.refStructures.push(refInfo);
            }));
        }
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class StructureReferTypeConvertor {
    /**
     * @param {?} obj
     * @param {?} datatype
     * @param {?} ctxt
     * @return {?}
     */
    convertToObject(obj, datatype, ctxt) {
        /** @type {?} */
        var referType = (/** @type {?} */ (datatype));
        obj["kind"] = referType.getKind();
        obj['referId'] = referType.referId;
        obj['code'] = referType.code;
        obj['name'] = referType.name;
    }
    /**
     * @param {?} dataType
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    initFromObject(dataType, obj, ctxt) {
        /** @type {?} */
        var refType = (/** @type {?} */ (dataType));
        refType.kind = obj['kind'];
        refType.referId = obj['referId'];
        refType.code = obj['code'];
        refType.name = obj['name'];
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DefaultGspComponentDataTypeSerializer {
    /**
     * @param {?} type
     * @param {?} ctxt
     * @return {?}
     */
    serialize(type, ctxt) {
        /** @type {?} */
        let result = new Object();
        /** @type {?} */
        let baseConvertor = new StructureReferTypeConvertor();
        baseConvertor.convertToObject(result, type, ctxt);
        return result;
    }
    /**
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    deserialize(obj, ctxt) {
        /** @type {?} */
        let result = new StructureReferType();
        /** @type {?} */
        let baseConvertor = new StructureReferTypeConvertor();
        baseConvertor.initFromObject(result, obj, ctxt);
        return result;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DefaultGspComponentDataTypeSerializerFactory {
    /**
     * @param {?} kind
     * @return {?}
     */
    get(kind) {
        switch (kind) {
            case GspComponentConstant.CMP_STRUCTURED_TYPE_REF:
                if (!this.serializer) {
                    this.serializer = new DefaultGspComponentDataTypeSerializer();
                }
                return this.serializer;
            default:
                return new DefaultDataTypeSerializerFactory().get(kind);
        }
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DefaultGspComponentSerializeContext {
    // private gspComponentFactory: IGspComponentSerializerFactory;
    /**
     * @return {?}
     */
    getDataTypeSerializerFactory() {
        if (!this.dataTypeFactory) {
            this.dataTypeFactory = new DefaultGspComponentDataTypeSerializerFactory();
        }
        return this.dataTypeFactory;
    }
    /**
     * @return {?}
     */
    getOperationSerializerFactory() {
        if (!this.operationFactory) {
            this.operationFactory = new DefaultOperationSerializerFactory();
        }
        return this.operationFactory;
    }
    /**
     * @return {?}
     */
    getParameterSerializerFactory() {
        if (!this.parameterFactory) {
            this.parameterFactory = new DefaultParameterSerializerFactory();
        }
        return this.parameterFactory;
    }
    /**
     * @return {?}
     */
    getPropertySerializerFactory() {
        if (!this.propertyFactory) {
            this.propertyFactory = new DefaultPropertySerializerFactory();
        }
        return this.propertyFactory;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DefaultGspComponentSerializer {
    /**
     * @param {?} component
     * @param {?} ctxt
     * @return {?}
     */
    serialize(component, ctxt) {
        /** @type {?} */
        let result = new Object();
        //使用基本的序列化器
        /** @type {?} */
        let baseConvertor = new GspComponentConvertor();
        baseConvertor.convertToObject(result, component, ctxt);
        //Cando:增加自己对Object的扩展
        return result;
    }
    /**
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    deserialize(obj, ctxt) {
        /** @type {?} */
        let result = new GspComponent();
        /** @type {?} */
        let baseConvertor = new GspComponentConvertor();
        baseConvertor.initFromObject(result, obj, ctxt);
        return result;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DefaultGspComponentSerializerFactory {
    constructor() {
        this.serializerDict = {};
    }
    /**
     * @param {?} kind
     * @return {?}
     */
    get(kind) {
        switch (kind) {
            case "Default":
                if (!this.serializerDict[kind]) {
                    /** @type {?} */
                    var serializer = new DefaultGspComponentSerializer();
                    this.serializerDict[kind] = serializer;
                }
                break;
            default:
                throw new Error(`未匹配到类型为${kind}的序列化器!`);
        }
        return this.serializerDict[kind];
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class CommonComponentService {
    /**
     * @param {?} http
     * @param {?} structureSvc
     */
    constructor(http, structureSvc) {
        this.http = http;
        this.structureSvc = structureSvc;
        this.commom_component_url = '/api/runtime/sys/v1.0/common-component';
    }
    /**
     * @param {?} componentId
     * @param {?} headers
     * @return {?}
     */
    getComponent(componentId, headers) {
        /** @type {?} */
        let url = this.commom_component_url + '?componentId=' + componentId;
        return this.http.get(url, { headers: headers }).pipe(map((/**
         * @param {?} cmpObj
         * @return {?}
         */
        cmpObj => {
            /** @type {?} */
            let serializer = new DefaultGspComponentSerializer();
            /** @type {?} */
            let ctxt = new DefaultGspComponentSerializeContext();
            /** @type {?} */
            let cmp = serializer.deserialize(cmpObj, ctxt);
            return cmp;
        })));
    }
    /**
     * 获取构件中操作参数的Schema
     * @param {?} componentId 构件Id
     * @param {?} operationCode 操作编号
     * @param {?} paramCode 参数编号
     * @param {?} headers 请求header，应包含sessionId等必需参数
     * @return {?}
     */
    getParamSchema(componentId, operationCode, paramCode, headers) {
        /** @type {?} */
        let url = this.commom_component_url + '?componentId=' + componentId;
        return this.http.get(url, { headers: headers }).pipe(mergeMap((/**
         * @param {?} cmpObj
         * @return {?}
         */
        cmpObj => {
            /** @type {?} */
            let serializer = new DefaultGspComponentSerializer();
            /** @type {?} */
            let ctxt = new DefaultGspComponentSerializeContext();
            /** @type {?} */
            let cmp = serializer.deserialize(cmpObj, ctxt);
            /** @type {?} */
            let op = cmp.operations.find((/**
             * @param {?} x
             * @return {?}
             */
            x => x.code == operationCode));
            if (op == null) {
                throw new Error(`CommonComponent Operation NotFound，OperationCode: ${operationCode}`);
            }
            /** @type {?} */
            let param = op.parameters.find((/**
             * @param {?} x
             * @return {?}
             */
            x => x.code == paramCode));
            if (param == null) {
                throw new Error(`CommonComponent Parameter NotFound，ParameterCode: ${paramCode}`);
            }
            /** @type {?} */
            let paramType = param.parameterType;
            //引用类型
            if (paramType instanceof StructureReferType) {
                return this.getStructureReferType(cmp, paramType, headers);
            }
            //列表类型
            else if (paramType instanceof CollectionType) {
                return this.getCollectionType(cmp, paramType, headers);
            }
            //字典类型
            else if (paramType instanceof MapType) {
                return this.getMapType(cmp, paramType, headers);
            }
            //基本类型
            else {
                return of(paramType);
            }
        })));
    }
    ;
    /**
     * 获取构件中操作返回值的Schema
     * @param {?} componentId 构件Id
     * @param {?} operationCode 操作编号
     * @param {?} headers 请求header，应包含sessionId等必需参数
     * @return {?}
     */
    getReturnValueSchema(componentId, operationCode, headers) {
        /** @type {?} */
        let url = this.commom_component_url + '?componentId=' + componentId;
        return this.http.get(url, { headers: headers }).pipe(mergeMap((/**
         * @param {?} cmpObj
         * @return {?}
         */
        cmpObj => {
            /** @type {?} */
            let serializer = new DefaultGspComponentSerializer();
            /** @type {?} */
            let ctxt = new DefaultGspComponentSerializeContext();
            /** @type {?} */
            let cmp = serializer.deserialize(cmpObj, ctxt);
            /** @type {?} */
            let op = cmp.operations.find((/**
             * @param {?} x
             * @return {?}
             */
            x => x.code == operationCode));
            if (op == null) {
                throw new Error(`CommonComponent Operation NotFound，OperationCode: ${operationCode}`);
            }
            /** @type {?} */
            let returnType = op.returnType;
            //引用类型
            if (returnType instanceof StructureReferType) {
                return this.getStructureReferType(cmp, returnType, headers);
            }
            //列表类型
            else if (returnType instanceof CollectionType) {
                return this.getCollectionType(cmp, returnType, headers);
            }
            //字典类型
            else if (returnType instanceof MapType) {
                return this.getMapType(cmp, returnType, headers);
            }
            //基本类型
            else {
                return of(returnType);
            }
        })));
    }
    /**
     * @private
     * @param {?} cmp
     * @param {?} dataType
     * @param {?} headers
     * @return {?}
     */
    getCollectionType(cmp, dataType, headers) {
        /** @type {?} */
        let elementType = dataType.elementType;
        if (elementType instanceof StructureReferType) {
            return this.getStructureReferType(cmp, elementType, headers).pipe(map((/**
             * @param {?} structuredType
             * @return {?}
             */
            structuredType => {
                dataType.elementType = structuredType;
                return dataType;
            })));
        }
        else if (elementType instanceof PrimitiveType) {
            return of(dataType);
        }
        else {
            throw new Error(`不支持的列表项类型: ${elementType.getKind()}`);
        }
    }
    /**
     * @private
     * @param {?} cmp
     * @param {?} dataType
     * @param {?} headers
     * @return {?}
     */
    getMapType(cmp, dataType, headers) {
        /** @type {?} */
        let valueType = dataType.valueType;
        if (valueType instanceof StructureReferType) {
            return this.getStructureReferType(cmp, valueType, headers).pipe(map((/**
             * @param {?} structuredType
             * @return {?}
             */
            structuredType => {
                dataType.valueType = structuredType;
                return dataType;
            })));
        }
        else if (valueType instanceof PrimitiveType) {
            return of(dataType);
        }
        else if (valueType instanceof CollectionType) {
            return this.getCollectionType(cmp, valueType, headers).pipe(map((/**
             * @param {?} collectionType
             * @return {?}
             */
            collectionType => {
                dataType.valueType = collectionType;
                return dataType;
            })));
        }
        else {
            throw new Error(`不支持的列表项类型: ${valueType.getKind()}`);
        }
    }
    /**
     * @private
     * @param {?} cmp
     * @param {?} referType
     * @param {?} headers
     * @return {?}
     */
    getStructureReferType(cmp, referType, headers) {
        /** @type {?} */
        let refInfo = cmp.refStructures.find((/**
         * @param {?} x
         * @return {?}
         */
        x => x.id == referType.referId));
        return this.structureSvc.get(refInfo.referType, refInfo.referId, headers).pipe(map((/**
         * @param {?} structureObj
         * @return {?}
         */
        structureObj => {
            //使用默认serializeContext进行反序列化
            /** @type {?} */
            let serializeCtxt = new DefaultSerializeContext();
            /** @type {?} */
            let kind = structureObj['kind'];
            /** @type {?} */
            let serializer = serializeCtxt.getStructureSerializerFactory().get(kind);
            /** @type {?} */
            let structure = (/** @type {?} */ (serializer.deserialize(structureObj, serializeCtxt)));
            /** @type {?} */
            let structuredTypeId = refInfo.structuredTypeId;
            /** @type {?} */
            let allStructures = this.getAllStructures(structure);
            /** @type {?} */
            let referType = allStructures.find((/**
             * @param {?} x
             * @return {?}
             */
            x => x.id == structuredTypeId));
            return referType;
        })));
    }
    /**
     * @private
     * @param {?} structure
     * @return {?}
     */
    getAllStructures(structure) {
        /** @type {?} */
        let array = new Array();
        if (structure.structuredTypes != null) {
            array = array.concat(structure.structuredTypes);
        }
        if (structure.refStructures != null) {
            structure.refStructures.forEach((/**
             * @param {?} refStructure
             * @return {?}
             */
            refStructure => {
                if (refStructure.structuredTypes != null) {
                    array = array.concat(refStructure.structuredTypes);
                }
            }));
        }
        return array;
    }
}
CommonComponentService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CommonComponentService.ctorParameters = () => [
    { type: HttpClient },
    { type: CommonStructureDTService }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class CommonComponentModule {
}
CommonComponentModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    HttpClientModule
                ],
                declarations: [],
                providers: [CommonComponentService],
                exports: []
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { GspComponent, StructureReferType, RefStructureInfo, GspComponentConstant, GspComponentConvertor, RefStructureUtil, StructureReferTypeConvertor, DefaultGspComponentSerializeContext, DefaultGspComponentDataTypeSerializerFactory, DefaultGspComponentSerializerFactory, DefaultGspComponentDataTypeSerializer, DefaultGspComponentSerializer, CommonComponentModule, CommonComponentService };
