// @farris/mobile-command-services/esm2015/lib/js-bridge-service/ccwork-service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/js-bridge-service/ccwork-service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
class CCworkService {
    /**
     * @param {?} callback
     * @return {?}
     */
    init(callback) {
        window.top['iCityBridge'].init(callback);
    }
    /**
     * {
     * quality:图片质量，可选值high、mid、low, high为1920*1080；mid为720*540；low: 480*360，默认为mid
     * mode:唤起模式，可选值both（仅支持iOS）、camera、album, 默认为camera;
     * mark文字水印，字数最大值50
     * }
     * @param {?=} options Object
     *
     * @return {?}
     */
    takePhoto(options = { quality: 'mid' }) {
        // callback return object {errormessage:""，result:"base64格式"}
        /** @type {?} */
        const subject = new Subject();
        /** @type {?} */
        const callback = (/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            if (data) {
                subject.next(data['result']);
            }
        });
        window.top['iCityBridge'].ccworkGetPicsFromCameraAndAlbum(Object.assign({}, options), callback);
        return subject;
    }
    /**
     * path 如果不为空则从指定图片文件识别二维码,空就吊起相机
     * filters 条码类型过滤列表，如QR、EAN13、EAN8
     *
     * @return {?}
     */
    scanQRCode() {
        /** @type {?} */
        const subject = new Subject();
        /** @type {?} */
        const callback = (/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            if (data) {
                subject.next(data['result']);
            }
        });
        window.top['iCityBridge'].ccworkScanCode({}, callback);
        return subject;
    }
    /**
     * @return {?}
     */
    getSysInfo() {
    }
    /**
     * @return {?}
     */
    getLocation() {
        /**
         *  回调函数
         *  输出结构 object
         *  errormessage:""
         *  result:Object {
         *  addr:'山东省济南市历下区经十路靠近浪潮'
         *  city:"济南"
         *  country:"中国"
         *  course:-1
         *  district:"历下区"
         *  lat:36.66254123263889
         *  lng:117.12974745008681
         *  province:"山东省"
         *  street:"经十路"
         *  streetNum:"1单元"
         * }
         *  status:true
         * @type {?}
         */
        const subject = new Subject();
        /** @type {?} */
        const callback = (/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            if (data) {
                subject.next(data['result']);
            }
        });
        window.top['iCityBridge'].ccworkGetLocation(callback);
        return subject;
    }
    // TODO:云上协同选择文件
    /**
     * @return {?}
     */
    selectFile() {
        /** @type {?} */
        const subject = new Subject();
        subject.next([]);
        return subject;
    }
    /**
     *
     * @param {?} number string
     * @param {?} content string
     * @return {?}
     */
    sendSMS(number, content) {
        /** @type {?} */
        const options = {
            numbers: [number],
            content
        };
        window.top['iCityBridge'].ccworkSendSms(options);
    }
    /**
     * @param {?} number
     * @param {?} content
     * @return {?}
     */
    openSendSMS(number, content) {
        window.top['iCityBridge'].sendSMS({ tel: number, msg: content });
    }
    /**
     *
     * @param {?} number string
     * @return {?}
     */
    tel(number) {
        window.top['iCityBridge'].ringUp(number);
    }
    /**
     *
     * @param {?=} options
     * @return {?}
     */
    selectPictureFromAlbum(options = { maxNum: 9, mType: 0 }) {
        /**
         * params object
         *  { maxNum : number 9 最大可选数 默认9
         *    mType: number 0 	选择模式，可选值0/1, 0:图片，1：视频;
         * }
         * @type {?}
         */
        const subject = new Subject();
        /** @type {?} */
        const callback = (/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            if (data) {
                subject.next(data['result']);
            }
        });
        window.top['iCityBridge'].ccworkGetPicsFromAlbum(options, callback);
        return subject;
        /**
         *  object
         *  errormessage:""
         *  result:Array [
         * base64
         * ]
         *  status:true
         */
    }
    /**
     *
     * @param {?} options
     * @return {?}
     */
    takeVideo(options) {
        /**
         * params object
         *  {
         *  time : number 短视频时间（秒） 最大60s
         *  quality: number 0 	视频质量 取值范围0~1
         *  id :string 	视频id（视频名字）
         * }
         * @type {?}
         */
        const subject = new Subject();
        /** @type {?} */
        const callback = (/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            if (data) {
                subject.next(data['result']);
            }
        });
        window.top['iCityBridge'].ccworkTakeShortVideo(Object.assign({ time: 15 }, options), callback);
        return subject;
        /**
         *  object
         *  errormessage:""
         *  result:Object {
         *  value:'path'
         * }
         *  status:true
         */
    }
    /**
     * @return {?}
     */
    closeWindow() {
        window.top['iCityBridge'].closeWebView();
    }
    /**
     *
     * @param {?} title string
     * @return {?}
     */
    setTitles(title) {
        window.top['iCityBridge'].setTitle(title);
    }
    /**
     * @param {?} options
     * @return {?}
     */
    downloadFile(options) {
        alert('暂不支持附件下载');
    }
    /**
     * @param {?=} fn
     * @return {?}
     */
    setOriginGoback(fn) {
        window.top['iCityBridge'].ccworkInterceptOriginGoback(!!fn, fn, (/**
         * @return {?}
         */
        () => { }));
    }
    /**
     * @param {?=} fn
     * @return {?}
     */
    pushOriginGoback(fn) {
        window['OriginGoback'] = window['OriginGoback'] || [];
        /** @type {?} */
        const OriginGoback = (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const gobackFn = window['OriginGoback'].pop();
            if (gobackFn) {
                gobackFn();
            }
            else {
                window.top['iCityBridge'].ccworkInterceptOriginGoback(false, (/**
                 * @return {?}
                 */
                () => { }), (/**
                 * @return {?}
                 */
                () => { }));
                history.back();
            }
        });
        if (fn) {
            window['OriginGoback'].push(fn);
            window.top['iCityBridge'].ccworkInterceptOriginGoback(true, OriginGoback, (/**
             * @return {?}
             */
            () => { }));
        }
    }
    /**
     * @param {?=} fn
     * @return {?}
     */
    popOriginGoback(fn) {
        window['OriginGoback'].pop();
    }
    /**
     * @param {?} opt
     * @return {?}
     */
    playShortVideo(opt) {
        window.top['iCityBridge'].ccworkPlayShortVideo(opt);
    }
}
export { CCworkService };
