// @qdp/localize/fesm5/qdp-localize.js
import { Injectable, LOCALE_ID, Inject, NgModule, defineInjectable, inject } from '@angular/core';
import { CommonUtils, FarrisCommonModule } from '@farris/ui-common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var TITLE_OPTION = {
    tab: 'Title',
    title: 'Title',
    subTitleColumnName: [
        'Type', 'Content', 'Level', 'Align'
    ],
    subTitleData: [
        { label: 'subtitle', value: 0 },
        { label: 'footer', value: 1 }
    ],
    subTitleAlign: [
        { label: 'left', value: 0 },
        { label: 'center', value: 1 },
        { label: 'right', value: 2 }
    ]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var COLUMN_OPTION = {
    tab: 'Column',
    btnAdd: 'Add',
    btnDelete: 'Delete',
    btnUp: 'MoveUp',
    btnDown: 'MoveDown',
    btnLeft: 'MoveLeft',
    btnRight: 'MoveRight',
    selection1: 'Basic Information',
    selection2: 'Advanced Information',
    name: 'Name',
    binding: 'Binding',
    showType: 'Display Type',
    colWidth: 'Column Width',
    align: 'Align',
    alignData: [
        { name: 'left', value: 0 },
        { name: 'center', value: 1 },
        { name: 'right', value: 2 }
    ],
    show: 'Show',
    fixColumn: 'Fix Column',
    wordWrap: 'Word Wrap',
    autoFitColumn: 'Auto Fit',
    dataShow: 'Data Show',
    formatType: 'Format Type',
    formatTypeData: [
        { name: 'select', value: 0 },
        { name: 'number', value: 1 },
        { name: 'date', value: 2 },
        { name: 'enum', value: 3 }
    ],
    dateFormat: 'Date Format',
    numberFormat: 'Number Format',
    numberField: 'Precision Description Field',
    enumFormat: 'Enum Format',
    sort: 'Sort',
    sortData: [
        { name: 'select', value: 0 },
        { name: 'asc', value: 1 },
        { name: 'desc', value: 2 }
    ],
    group: 'Grouping',
    pageTotal: 'Subtotal of This Page Text',
    pageTotalType: 'Subtotal of This Page Total Type',
    groupTotal: 'Grouping Total Text',
    groupSubTotalType: 'Grouping subtotal Types',
    groupTotalType: 'Group total type',
    aggregateTotal: 'Aggregate Total Text',
    totalType: 'Total Type',
    aggregateData: [
        { name: 'select', value: 0 },
        { name: 'sum', value: 1 },
        { name: 'min', value: 2 },
        { name: 'max', value: 3 },
        { name: 'count', value: 4 },
        { name: 'average', value: 5 }
    ],
    dimension: 'Dimension',
    dimensionData: [
        { name: 'select', value: 0 },
        { name: 'row dimension', value: 1 },
        { name: 'col dimension', value: 2 },
        { name: 'value dimension', value: 3 }
    ],
    enumType: [
        { name: 'index', value: '0' },
        { name: 'code', value: '1' },
        { name: 'name', value: '2' }
    ],
    groupColor: 'Grouping Row Color',
    dataShowOption: [
        { name: 'Default', value: -1 },
        { name: 'Auto Fit', value: 1 },
        { name: 'Wrap', value: 2 }
    ],
    isRed: 'Less than zero displays red font'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var STYLE_OPTION = {
    tab: 'Style',
    column: 'Column',
    styleTitle: 'Title',
    styleSubTitle: 'SubTitle',
    styleHeader: 'Header',
    styleBody: 'DataArea',
    styleFooter: 'Footer',
    rowHeight: 'Row Height',
    textStyle: 'Content Style',
    preview: 'Preview',
    fontFamilyData: [
        { name: '宋体', value: 0 },
        { name: '新宋体', value: 1 },
        { name: '仿宋', value: 2 },
        { name: '楷体', value: 3 },
        { name: '黑体', value: 4 },
        { name: 'ARIAL', value: 5 },
        { name: 'ROMAN', value: 6 },
        { name: 'TIMES NEW ROMAN', value: 7 },
        { name: 'VERDANA', value: 8 },
        { name: 'Microsoft YaHei', value: 9 }
    ]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var OTHER_OPTION = {
    tab: 'Other',
    selection: 'Basic Information',
    dataCache: 'Cache',
    firstLoad: 'First Query',
    filterRow: 'Enable the filter row',
    showHeaderSort: 'Enable column header sorting',
    mergeCell: 'Enable merge cell',
    pageType: 'Pagination Type',
    pageSize: 'Pagination Size',
    delay: 'Data Delay Loading',
    timing: 'Data Timing Loading',
    export: 'Export Range',
    groupOption: 'Grouping Option',
    rowSubtotal: 'Row Subtotal',
    colSubtotal: 'Column Subtotal',
    rowTotal: 'Row Aggregate',
    colTotal: 'Column Aggregate',
    showZero: '0 Value Displayed',
    expand: 'Expand by Default ',
    pageTypeData: [
        { name: 'all loading', value: 0 },
        { name: 'paging loading', value: 1 },
    ],
    rangeData: [
        { name: 'all', value: 0 },
        { name: 'current page', value: 1 },
        { name: 'custom', value: 2 }
    ],
    treeInfo: 'TreeList Option',
    treeInfoType: 'Construction Mode',
    treeInfoData: [
        { name: 'Select', value: 2 },
        { name: 'Path Code', value: 0 },
        { name: 'Autocorrelation', value: 1 }
    ],
    pathField: 'Path Field',
    pathStep: 'Path Step',
    idField: 'ID Field',
    parentField: 'Parent Field',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var SCHEME_MANAGEMENT = {
    dailogName: 'Format Scheme',
    formatSchemaManagerName: 'Format Scheme Management',
    formatManager: 'Format Management',
    preFormat: 'Prefabrication Scheme',
    btnCancel: 'Cancel',
    btnOK: 'OK',
    btnSave: 'Save',
    btnSaveAs: 'Save As',
    btnDelete: 'Delete',
    btnSaveAsClose: 'Save And Close',
    schemeName: 'Name',
    schemeType: 'Type',
    schemeTypePublic: 'Public',
    schemePublic: 'Public Scheme',
    schemeTypePersonal: 'Personal',
    schemePersonal: 'Personal Scheme',
    schemeTypeOrg: 'Department',
    schemeOrg: 'Department Scheme',
    schemeDefault: 'Default',
    titleOption: TITLE_OPTION,
    columnOption: COLUMN_OPTION,
    styleOption: STYLE_OPTION,
    otherOption: OTHER_OPTION,
    message: {
        save: {
            success: 'save success !',
            error: 'save error!',
        },
        update: {
            success: 'update success !',
            error: 'update error!',
        },
        delete: {
            success: 'delete success !',
            error: 'delete error!',
        },
        info: {
            title: 'info',
            delete: 'Please select the scheme to delete!'
        },
        validate: {
            judgeUpdate: 'The format scheme has been modified, whether to abandon the modification?',
            notNull: 'The scheme name should not be empty!',
            exists: 'The scheme name already exists!',
            tooLong: 'The scheme name must not exceed 100 characters!',
            isProtected: 'System prefabrication cannot be deleted or modified!',
            crosstabValueDimension: 'Value dimensions must be numeric columns!',
            qdpModifyPublicSchema: 'No permission to operate the public scheme!',
            textColumnSort: 'column type are text and do not support sorting operations!'
        }
    },
    basicInfoName: 'Basic Information'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CHARTS_TYPE_OPTION = {
    tab: 'Charts Type',
    chartsType: 'Charts Type',
    chartsTypeData: [
        { name: 'select', value: -1 },
        { name: 'line', value: 'line' },
        { name: 'bar', value: 'bar' },
        { name: 'pie', value: 'pie' },
        { name: 'area', value: 'line1' },
        { name: 'scatter', value: 'scatter' },
        { name: 'effectScatter', value: 'effectScatter' }
    ]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CHARTS_DATA_OPTION = {
    tab: 'Charts Data',
    xAxis: 'xAxis',
    xPositionData: [
        { name: 'bottom', value: 'bottom' },
        { name: 'top', value: 'top' }
    ],
    yAxis: 'yAxis',
    yPositionData: [
        { name: 'left', value: 'left' },
        { name: 'right', value: 'right' }
    ],
    axisTypeData: [
        { name: 'category', value: 'category' },
        { name: 'value', value: 'value' }
    ],
    type: 'Type',
    position: 'Position',
    binding: 'Binding',
    category: 'Category',
    numerical1: 'Numerical1',
    numerical2: 'Numerical2',
    // 经度
    longitude: 'Longitude',
    // 纬度
    latitude: 'Latitude',
    barTypeData: [
        { name: 'single column', value: 1 },
        { name: 'multiple columns', value: 2 },
        { name: 'superposition', value: 3 },
    ],
    name: 'Name',
    select: 'Select',
    dimensionSetting: 'Dimension Option',
    dimension: 'Dimension',
    indicator: 'Indicator'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CHARTS_STYLE_OPTION = {
    tab: 'Charts Style',
    titleOption: 'Title Option',
    title: 'Title',
    select: 'Select',
    xAxisOption: 'XAxis Option',
    xAxisPosition: 'xAxis Position',
    yAxisPosition: 'yAxis Position',
    xAxisAlignData: [
        { name: 'left', value: 'left' },
        { name: 'center', value: 'center' },
        { name: 'right', value: 'right' }
    ],
    yAxisOption: 'YAxis Option',
    yAxisAlignData: [
        { name: 'top', value: 'top' },
        { name: 'middle', value: 'middle' },
        { name: 'bottom', value: 'bottom' }
    ],
    fontFamilyData: [
        { name: 'Microsoft YaHei', value: 'Microsoft YaHei' },
        { name: 'sans-serif', value: 'sans-serif' },
        { name: 'serif', value: 'serif' },
        { name: 'monospace', value: 'monospace' },
        { name: 'Arial', value: 'Arial' },
        { name: 'Courier New', value: 'Courier New' }
    ],
    rowHeight: 'Row Height',
    contentStyle: 'Content Otyle',
    multiDataOption: 'Multidimensional Data Option',
    dataMultiple: [
        { name: 'side', value: 'side' },
        { name: 'pile', value: 'pile' }
    ],
    multiDataType: 'Type',
    legend: 'Legend',
    legendAlgin: 'Align',
    legendAlignData: [
        { name: 'auto', value: 'auto' },
        { name: 'left', value: 'left' },
        { name: 'right', value: 'right' }
    ],
    legendOrient: 'Orient',
    orientData: [
        { name: 'horizontal', value: 'horizontal' },
        { name: 'vertical', value: 'vertical' }
    ],
    legendGap: 'Gap',
    itemWidth: 'Item Width',
    itemHeight: 'Item Height',
    visualMap: 'Visualmap',
    visualMapType: 'Type',
    visualMapTypeData: [
        { name: 'continuous', value: 'continuous' },
        { name: 'piecewise', value: 'piecewise' }
    ],
    visualMapOrient: 'Orient',
    visualMapPosition: 'Position',
    visualMapTop: 'Top',
    visualMapBottom: 'Bottom',
    visualMapLeft: 'Left',
    visualMapRight: 'Right',
    continuous: 'Continuous',
    piecewise: 'Piecewise',
    otherStyle: 'Other',
    customStyle: 'Custom Style',
    validate: {
        echartsCustomOptions: 'Illegal Json object format!'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CHARTS_SCHEME_MANAGEMENT = {
    chartsSchemaManagerName: 'Charts Scheme Management',
    chartsType: CHARTS_TYPE_OPTION,
    chartsData: CHARTS_DATA_OPTION,
    chartsStyle: CHARTS_STYLE_OPTION
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var PARAMS_MAPPING_OPTION = {
    paramOption: 'Parameters Option',
    btnAdd: 'Add',
    btnDelete: 'Delete',
    btnCancel: 'Cancel',
    btnOK: 'OK',
    columnHeader: [
        'Parameter', 'Mapping', 'Source Parameter', 'Constant', 'Expression'
    ],
    mappingTypeData: [
        { name: 'field', value: 0 },
        { name: 'constant', value: 1 },
        { name: 'expression', value: 2 }
    ],
    message: {
        validate: {
            select: 'please select a record to delete!'
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var DRILL_DOWN_OPTION = {
    drillDownName: 'Drill Down Management',
    btnAdd: 'Add',
    btnDelete: 'Delete',
    btnCancel: 'Cancel',
    btnOK: 'OK',
    menuBtn: 'Menu',
    appBtn: 'App',
    columnHeader: [
        'Function Code', 'Title', 'Drill down primary key', 'Expend expression', 'Presentation form', 'Double Click Drill down', 'Link Field'
    ],
    paramColumnHeader: [
        'Parameter', 'Name', 'Value', 'Is Expression'
    ],
    displayStyle: [
        { 'name': 'Right click associated query', value: 1 },
        { 'name': 'Hyperlink query', value: 2 }
    ],
    enumExpress: [
        'Yes', 'No'
    ]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CONDITION_SCHEMA_MANAGER = {
    conditionSchemeManagerName: 'Condition Scheme Management',
    schemeSave: 'Scheme Save',
    schemeManagement: 'Scheme Management',
    btnDelete: 'Delete',
    btnSave: 'Save',
    schemeName: 'Scheme Name',
    schemeType: 'Scheme Type',
    saveScheme: 'Save Scheme',
    public: 'Public',
    personal: 'Personal',
    default: 'Default',
    noDefault: 'Not Default',
    message: {
        save: {
            success: 'save success!',
            error: 'save error!'
        },
        update: {
            success: 'update success!',
            error: 'update error!'
        },
        delete: {
            success: 'delete success!',
            error: 'delete error!'
        },
        validate: {
            delete: 'the default scheme does not deletion!',
            longLimit: 'name are too long!',
            repeatName: 'duplicate name!',
            select: 'please select a record to delete!'
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var PRINT_INTEGRATION_ZH_CHS = {
    print: 'Print',
    dailogName: 'Print Option',
    layout: 'Layout',
    paperSize: 'Paper Size',
    margin: 'Margin',
    zoom: 'Zoom',
    pdfDownload: 'PDF Download',
    printPreview: 'Print Preview',
    distributionData: [
        { name: 'horizontal', value: 1 },
        { name: 'vertical', value: 0 }
    ],
    pageSizeData: [
        { name: 'Letter paper', value: 1 },
        { name: 'Tabloid', value: 3 },
        { name: 'Legal paper', value: 5 },
        { name: 'Statement', value: 6 },
        { name: 'A3', value: 8 },
        { name: 'A4', value: 9 },
        { name: 'A5', value: 11 },
        { name: 'B4 (JIS)', value: 12 },
        { name: 'B5 (JIS)', value: 13 }
    ],
    marginData: [
        { name: 'default', value: 0, margin: { marginTop: 1, marginLeft: 1, marginBottom: 1, marginRight: 1 } },
        { name: 'none', value: 1, margin: { marginTop: 0, marginLeft: 0, marginBottom: 0, marginRight: 0 } },
        { name: 'min', value: 2, margin: { marginTop: 0, marginLeft: 0, marginBottom: 0, marginRight: 0 } },
        { name: 'custom', value: 3, margin: { marginTop: 1, marginLeft: 1, marginBottom: 1, marginRight: 1 } }
    ],
    message: {
        validate: {
            margin: 'Please select the margins!',
            pageSize: 'Please select paper!',
            distribution: 'Please select the printing direction!'
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var IDE_CMP = {
    schemaManager: SCHEME_MANAGEMENT,
    chartsSchemaManager: CHARTS_SCHEME_MANAGEMENT,
    conditionSchemaManager: CONDITION_SCHEMA_MANAGER,
    paramsMapping: PARAMS_MAPPING_OPTION,
    drillDown: DRILL_DOWN_OPTION,
    printOption: PRINT_INTEGRATION_ZH_CHS
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var QUERY_FRAMEWORK = {
    exportName: 'Export',
    btnExportText: 'Excel(Text)',
    btnExportValue: 'Excel(Number)',
    pageInfo: {
        totalPreInfo: 'Total',
        totalEndInfo: 'Record'
    },
    printName: 'Print',
    refreshName: 'Refresh',
    filterName: 'Filter',
    dimensionSetting: 'Dimension Setting',
    btnCondition: 'Query Condition',
    btnQuery: 'Query',
    btnSaveAsSchema: 'Save As',
    btnSaveSchema: 'Save',
    btnPrintForm: 'Documents Print',
    conditionManagerListColumn: ['Schema', 'Default', 'Operation'],
    btnOk: 'OK',
    btnCancel: 'Cancel',
    dialogTitle: 'Query Condition',
    dialogTitle1: 'Schema Management',
    dialogTitle2: 'Save As',
    tabName1: 'Commonly Used Condition',
    tabName2: 'Advanced Query',
    schemaName: 'Schema Name',
    ispreName: 'System prefabricated',
    defaultSchemaName: 'isDefault',
    schemaType: {
        public: 'Public',
        personal: 'Personal'
    },
    defaultName: 'Default',
    message: {
        saveas: 'The scheme name already exists, please modify it!',
        lengthValidate: 'Schema names are no more than 15 characters!',
        save: {
            success: 'save success !',
            error: 'save error!',
        },
        update: {
            success: 'update success !',
            error: 'update error!',
        },
        delete: {
            success: 'delete success !',
            error: 'delete error!',
        },
        info: {
            title: 'info',
            delete: 'Please select the scheme to delete!'
        },
        operate: {
            success: 'operate succes!',
            error: 'operate error!',
        },
        validate: {
            judgeUpdate: 'The format scheme has been modified, whether to abandon the modification?',
            notNull: 'The scheme name should not be empty!',
            exists: 'The scheme name already exists!',
            beMetadataIdEmpty: 'The BE metadata ID corresponding to the form printing is empty!',
            formIdFieldEmpty: 'The primary key field of the form printing form is empty!',
            unselected: 'Please select the data to print!'
        }
    },
    clearCache: 'Clear Cache',
    btnExpand: 'Expand All',
    btnCollapse: 'Collapse All',
    btnClearSort: 'Clear Sort'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var SPREAD = {
    crossTable: {
        dimensionManager: 'Dimension Setting',
        dimensionName: [
            'unselect', 'row dimension', 'column dimension', 'value dimension'
        ],
        columnManager: 'Column Setting',
        columnSetting: [
            'unselect', 'column setting', 'column group'
        ],
        valueDimensionTitle: 'Please drag the numeric field here',
        rowDimensionTitle: 'Please drag the text field here',
        colDimensionTitle: 'Please drag the text field here',
    },
    aggreateText: 'summary',
    message: {
        export: 'The system does not support the export of prefabricated scheme. Please save the prefabricated scheme before executing the export operation!',
        print: 'The system does not support the print of prefabricated scheme. Please save the prefabricated scheme before executing the print operation!',
        preview: 'The system does not support the preview of prefabricated scheme. Please save the prefabricated scheme before executing the preview operation!',
        aysnExport: 'Due to the large amount of data, the system carries out asynchronous export. After the export is completed, please check it in the system message！',
        clearCache: 'Cache cleared successfully!',
        extend: {
            query: {
                before: 'before query',
                after: 'after query',
            },
            dataLoading: {
                before: 'before data loading',
                after: 'after data loading'
            },
            render: {
                before: 'before render',
                after: 'after render'
            }
        },
        click: {
            select: 'selected event',
            cellSelect: 'selected cell event',
        },
        validate: {
            schemaInfo: 'Maintain the format scheme for the current query!',
            rowDimension: 'Maintain the row dimension of the corresponding format scheme for the current query!',
            valueDimension: 'Maintain the value dimension of the corresponding format scheme for the current query',
            nodata: 'No data',
            sort: 'Unable to modify the system preform, please save the scheme first!',
            qdpQueryExport: 'No permission to export!',
            qdpQueryPrint: 'No permission to print!',
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var Formular = {
    define: {
        labelSaveFormular: 'SaveFormula',
        labelSaveAsFormular: 'SaveAsFormula',
        labelDeleteFormular: 'DeleteFormula',
        labelGeneraterFormular: 'GenerateFormula',
        labelQueryFunc: 'QueryFunc',
        labelBasicInfo: 'BasicInfo',
        labelFormularName: 'FormulaName',
        labelIsGroup: 'IsGroup',
        labelParamInfo: 'ParamInfo',
        labelResultColInfo: 'ResultColInfo',
        resultSortColumns: [
            {
                id: '0',
                label: 'None'
            },
            {
                id: '1',
                label: 'Asc'
            },
            {
                id: '2',
                label: 'Desc'
            }
        ],
        resultColVisibleColumns: [
            {
                id: '0',
                label: 'False'
            },
            {
                id: '1',
                label: 'True'
            }
        ],
        columntitles: {
            'resultcolvisible': 'isQuery',
            'resultcode': 'resultCol',
            'resultdescription': 'description',
            'resultsort': 'resultSort'
        },
        messageAlert: {
            'nameVerify': 'FormulaName required！',
            'funcVerify': 'QueryFunc required！',
            'saveFormular': 'Formula save successed',
            'saveAsFormular': 'Formula save as successed',
            'saveAsFormularFailture': 'Formula save as failture',
            'pleaseSelectQueryFunc': 'Please select QueryFunc！',
            'pleaseSelectFormular': 'Please select Formular！',
            'resultVerify': 'Keep one column in the resultInfo as the return value！',
            'formularResultVerify': 'Keep at least one column in the resultInfo as the return value！',
            'combineVerify': 'Please maintain the composition formula！',
            'formularIntegrityVerify': 'Formula is not correct, please check and click finish!',
            'saveFormularFailure': 'Save formula failed！',
            'formularSaveVerify': 'Please save the formula and click finish！',
            'deleteFormular': 'Delete formula successed!',
            'deleteFormularFailure': 'Delete formula failed!',
            'formularIsProtected': 'This formular is protected and cannot be modified.',
            'formularIsProtectedNotDelete': 'This formular is protected and cannot be deleted.'
        }
    },
    defineWizard: {
        labelSelectTemplate: 'Select Wizard',
        labelFormularDefine: 'Formula Definition',
    },
    step: {
        labelPre: 'PreStep',
        labelNext: 'NextStep',
        labelFinish: 'Finish',
    },
    formularSelect: {
        selectionFormular: '[{"value":"0","name":"formula select"},{"value":"1","name":"formula create"},{"value":"2","name":"compose formula"}]',
        lableQdpFormular: 'QdpFormula',
        messageAlert: {
            'formularSelectVerify': 'Please select the formula！',
            'formularDefineVerify': 'Please define the take number formula or combination formula！',
        }
    },
    service: {
        lableFormulaWizard: 'Formula Wizard',
        messageAlert: {
            'formulaVerify': 'No valid formula is currently formed, confirm to close'
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var QDP_EN_US = {
    ideCmp: IDE_CMP,
    queryFramework: QUERY_FRAMEWORK,
    spread: SPREAD,
    formular: Formular
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var TITLE_OPTION$1 = {
    tab: '标题',
    title: '标题',
    subTitleColumnName: [
        '类型', '内容', '级次', '对齐'
    ],
    subTitleData: [
        { label: '副标题', value: 0 },
        { label: '表尾', value: 1 }
    ],
    subTitleAlign: [
        { label: '左', value: 0 },
        { label: '中', value: 1 },
        { label: '右', value: 2 }
    ]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var COLUMN_OPTION$1 = {
    tab: '列',
    btnAdd: '新增',
    btnDelete: '删除',
    btnUp: '上移',
    btnDown: '下移',
    btnLeft: '左移',
    btnRight: '右移',
    selection1: '基础信息',
    selection2: '高级信息',
    name: '名称',
    binding: '绑定字段',
    showType: '显示类型',
    colWidth: '列宽',
    align: '对齐方式',
    alignData: [
        { name: '左', value: 0 },
        { name: '中', value: 1 },
        { name: '右', value: 2 }
    ],
    show: '显示',
    fixColumn: '固定列',
    wordWrap: '折行',
    dataShow: '数据展示',
    autoFitColumn: '自适应',
    formatType: '格式化类型',
    formatTypeData: [
        { name: '请选择', value: 0 },
        { name: '数字', value: 1 },
        { name: '日期', value: 2 },
        { name: '逻辑', value: 3 }
    ],
    dateFormat: '日期格式',
    numberFormat: '数字格式',
    numberField: '精度描述字段',
    enumFormat: '格式串',
    sort: '排序',
    sortData: [
        { name: '请选择', value: 0 },
        { name: '正序', value: 1 },
        { name: '倒序', value: 2 }
    ],
    group: '是否分组',
    pageTotal: '本页小计文本',
    pageTotalType: '本页小计类型',
    groupTotal: '分组小计文本',
    groupSubTotalType: '分组小计类型',
    groupTotalType: '分组合计类型',
    aggregateTotal: '汇总合计文本',
    totalType: '汇总合计类型',
    aggregateData: [
        { name: '请选择', value: 0 },
        { name: '合计', value: 1 },
        { name: '最小值', value: 2 },
        { name: '最大值', value: 3 },
        { name: '计数', value: 4 },
        { name: '平均值', value: 5 }
    ],
    dimension: '交叉维度',
    dimensionData: [
        { name: '请选择', value: 0 },
        { name: '行维度', value: 1 },
        { name: '列维度', value: 2 },
        { name: '值维度', value: 3 }
    ],
    enumType: [
        { name: '索引', value: '0' },
        { name: '编号', value: '1' },
        { name: '名称', value: '2' }
    ],
    groupColor: '分组行颜色',
    dataShowOption: [
        { name: '默认', value: -1 },
        { name: '自适应', value: 1 },
        { name: '折行', value: 2 }
    ],
    isRed: '负数红字'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var STYLE_OPTION$1 = {
    tab: '样式',
    column: '栏目',
    styleTitle: '标题',
    styleSubTitle: '副标题',
    styleHeader: '表头',
    styleBody: '表体',
    styleFooter: '表尾',
    rowHeight: '行高',
    textStyle: '文本样式',
    preview: '预览',
    fontFamilyData: [
        { name: '宋体', value: 0 },
        { name: '新宋体', value: 1 },
        { name: '仿宋', value: 2 },
        { name: '楷体', value: 3 },
        { name: '黑体', value: 4 },
        { name: 'ARIAL', value: 5 },
        { name: 'ROMAN', value: 6 },
        { name: 'TIMES NEW ROMAN', value: 7 },
        { name: 'VERDANA', value: 8 },
        { name: 'Microsoft YaHei', value: 9 }
    ]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var OTHER_OPTION$1 = {
    tab: '其它',
    selection: '基本信息',
    dataCache: '数据缓存',
    firstLoad: '首次查询',
    filterRow: '开启过滤行',
    showHeaderSort: '开启列头排序',
    mergeCell: '是否合并单元格',
    pageType: '分页类型',
    pageSize: '分页大小',
    delay: '数据延迟加载',
    timing: '数据定时加载',
    export: '导出范围',
    groupOption: '分组设置',
    rowSubtotal: '行小计',
    colSubtotal: '列小计',
    rowTotal: '行合计',
    colTotal: '列合计',
    showZero: '0值是否显示',
    expand: '默认展开',
    pageTypeData: [
        { name: '全部加载', value: 0 },
        { name: '分页加载', value: 1 },
    ],
    rangeData: [
        { name: '全部', value: 0 },
        { name: '当前页', value: 1 },
        { name: '自定义', value: 2 }
    ],
    treeInfo: '树表设置',
    treeInfoType: '构建类型',
    treeInfoData: [
        { name: '请选择', value: 2 },
        { name: '分级码', value: 0 },
        { name: '自关联', value: 1 }
    ],
    pathField: '分级码字段',
    pathStep: '分级码长度',
    idField: '主键字段',
    parentField: '父级字段',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var SCHEME_MANAGEMENT$1 = {
    dailogName: '格式方案',
    formatSchemaManagerName: '格式方案管理',
    formatManager: '格式管理',
    preFormat: '系统预制格式',
    btnCancel: '取消',
    btnOK: '确定',
    btnSave: '保存',
    btnSaveAs: '另存为',
    btnDelete: '删除',
    btnSaveAsClose: '保存并关闭',
    schemeName: '名称',
    schemeType: '范围',
    schemeTypePublic: '公共',
    schemePublic: '公共方案',
    schemeTypePersonal: '个人',
    schemePersonal: '个人方案',
    schemeTypeOrg: '单位',
    schemeOrg: '单位方案',
    schemeDefault: '默认',
    titleOption: TITLE_OPTION$1,
    columnOption: COLUMN_OPTION$1,
    styleOption: STYLE_OPTION$1,
    otherOption: OTHER_OPTION$1,
    message: {
        save: {
            success: '保存成功！',
            error: '保存失败！',
        },
        update: {
            success: '更新成功！',
            error: '更新失败！',
        },
        delete: {
            success: '删除成功！',
            error: '删除失败！',
        },
        info: {
            title: '提示信息',
            delete: '请选择要删除的方案！'
        },
        validate: {
            judgeUpdate: '格式方案已修改, 是否放弃修改?',
            notNull: '方案名称不可以为空！',
            exists: '方案名称已存在！',
            tooLong: '方案名称不可以超过100个字符！',
            isProtected: '系统预制方案无法删除和修改！',
            crosstabValueDimension: '值维度必须是数值列！',
            qdpModifyPublicSchema: '没有操作公共方案的权限！',
            textColumnSort: '列是大文本，不支持排序操作！'
        }
    },
    basicInfoName: '基本信息'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CHARTS_TYPE_OPTION$1 = {
    tab: '类型',
    chartsType: '图表类型',
    chartsTypeData: [
        { name: '请选择', value: -1 },
        { name: '折线图', value: 'line' },
        { name: '柱状图', value: 'bar' },
        { name: '饼图', value: 'pie' },
        { name: '面积图', value: 'line1' },
        { name: '散点图', value: 'scatter' },
        { name: '地图散点', value: 'effectScatter' }
    ]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CHARTS_DATA_OPTION$1 = {
    tab: '数据',
    xAxis: 'x轴',
    xPositionData: [
        { name: '下', value: 'bottom' },
        { name: '上', value: 'top' }
    ],
    yAxis: 'y轴',
    yPositionData: [
        { name: '左', value: 'left' },
        { name: '右', value: 'right' }
    ],
    axisTypeData: [
        { name: '分类', value: 'category' },
        { name: '值', value: 'value' }
    ],
    type: '类型',
    position: '位置',
    binding: '绑定字段',
    category: '分类',
    numerical1: '数值1',
    numerical2: '数值2',
    // 经度
    longitude: '经度',
    // 纬度
    latitude: '纬度',
    barTypeData: [
        { name: '单列', value: 1 },
        { name: '多列', value: 2 },
        { name: '叠加', value: 3 },
    ],
    name: '名称',
    select: '请选择',
    dimensionSetting: '维度设置',
    dimension: '维度',
    indicator: '指标'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CHARTS_STYLE_OPTION$1 = {
    tab: '样式',
    titleOption: '标题设置',
    title: '标题',
    select: '请选择',
    xAxisOption: 'x轴设置',
    xAxisPosition: 'x轴位置',
    yAxisPosition: 'y轴位置',
    xAxisAlignData: [
        { name: '左', value: 'left' },
        { name: '中', value: 'center' },
        { name: '右', value: 'right' }
    ],
    yAxisOption: 'y轴设置',
    yAxisAlignData: [
        { name: '上', value: 'top' },
        { name: '中', value: 'middle' },
        { name: '下', value: 'bottom' }
    ],
    fontFamilyData: [
        { name: 'Microsoft YaHei', value: 'Microsoft YaHei' },
        { name: 'sans-serif', value: 'sans-serif' },
        { name: 'serif', value: 'serif' },
        { name: 'monospace', value: 'monospace' },
        { name: 'Arial', value: 'Arial' },
        { name: 'Courier New', value: 'Courier New' }
    ],
    rowHeight: '行高',
    contentStyle: '文本样式',
    multiDataOption: '多维数据设置',
    dataMultiple: [
        { name: '并排', value: 'side' },
        { name: '堆叠', value: 'pile' }
    ],
    multiDataType: '类型',
    legend: '图例',
    legendAlgin: '对齐方式',
    legendAlignData: [
        { name: '自适应', value: 'auto' },
        { name: '左', value: 'left' },
        { name: '右', value: 'right' }
    ],
    legendOrient: '布局方向',
    orientData: [
        { name: '水平', value: 'horizontal' },
        { name: '垂直', value: 'vertical' }
    ],
    legendGap: '间距',
    itemWidth: '宽度',
    itemHeight: '高度',
    visualMap: '视觉地图',
    visualMapType: '类型',
    visualMapTypeData: [
        { name: 'continuous', value: 'continuous' },
        { name: 'piecewise', value: 'piecewise' }
    ],
    visualMapOrient: '布局方向',
    visualMapPosition: '位置',
    visualMapTop: 'Top',
    visualMapBottom: 'Bottom',
    visualMapLeft: 'Left',
    visualMapRight: 'right',
    continuous: 'Continuous',
    piecewise: 'Piecewise',
    otherStyle: '其它',
    customStyle: '自定义',
    validate: {
        echartsCustomOptions: '非法的Json对象格式!'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CHARTS_SCHEME_MANAGEMENT$1 = {
    chartsSchemaManagerName: '格式方案管理',
    chartsType: CHARTS_TYPE_OPTION$1,
    chartsData: CHARTS_DATA_OPTION$1,
    chartsStyle: CHARTS_STYLE_OPTION$1
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var PARAMS_MAPPING_OPTION$1 = {
    paramOption: '参数设置',
    btnAdd: '新增',
    btnDelete: '删除',
    btnCancel: '取消',
    btnOK: '确定',
    columnHeader: [
        '参数名', '映射类别', '源参数', '常量', '表达式', '作为查询参数'
    ],
    mappingTypeData: [
        { name: '字段', value: 0 },
        { name: '常量', value: 1 },
        { name: '表达式', value: 2 }
    ],
    message: {
        validate: {
            select: '请选择一条要删除的记录!'
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var DRILL_DOWN_OPTION$1 = {
    drillDownName: '联查管理',
    btnAdd: '新增',
    btnDelete: '删除',
    btnCancel: '取消',
    btnOK: '确定',
    menuBtn: '菜单',
    appBtn: '应用',
    columnHeader: [
        '功能编号', '标题', '联查主键依据', '联查展现形式', '可见表达式', '双击联查', '超链字段'
    ],
    paramColumnHeader: [
        '参数', '名称', '值', '是否为表达式'
    ],
    displayStyle: [
        { 'name': '右键联查', value: 1 },
        { 'name': '超链联查', value: 2 }
    ],
    enumExpress: [
        '是', '否'
    ]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CONDITION_SCHEMA_MANAGER$1 = {
    conditionSchemeManagerName: '查询条件方案管理',
    schemeSave: '保存方案',
    schemeManagement: '方案管理',
    btnDelete: '删除',
    btnSave: '保存',
    schemeName: '方案名称',
    schemeType: '方案类别',
    saveScheme: '保存方案',
    public: '公共',
    personal: '个人',
    default: '默认',
    noDefault: '非默认',
    message: {
        save: {
            success: '保存成功!',
            error: '保存失败!'
        },
        update: {
            success: '修改默认方案成功!',
            error: '修改失败!请联系管理员!'
        },
        delete: {
            success: '删除成功!',
            error: '删除失败!'
        },
        validate: {
            delete: '默认方案不允许删除!',
            longLimit: '方案名称过长!',
            repeatName: '方案名称已存在!',
            select: '请选择一条要删除的记录!'
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var PRINT_INTEGRATION_ZH_CHS$1 = {
    print: '打印',
    dailogName: '打印设置',
    layout: '布局',
    paperSize: '纸张尺寸',
    margin: '边距',
    zoom: '缩放',
    pdfDownload: 'PDF下载',
    printPreview: '打印预览',
    distributionData: [
        { name: '横向', value: 1 },
        { name: '纵向', value: 0 }
    ],
    pageSizeData: [
        { name: '信纸', value: 1 },
        { name: 'Tabloid', value: 3 },
        { name: '法律专用纸', value: 5 },
        { name: 'Statement', value: 6 },
        { name: 'A3', value: 8 },
        { name: 'A4', value: 9 },
        { name: 'A5', value: 11 },
        { name: 'B4 (JIS)', value: 12 },
        { name: 'B5 (JIS)', value: 13 }
    ],
    marginData: [
        { name: '默认', value: 0, margin: { marginTop: 1, marginLeft: 1, marginBottom: 1, marginRight: 1 } },
        { name: '无', value: 1, margin: { marginTop: 0, marginLeft: 0, marginBottom: 0, marginRight: 0 } },
        { name: '最小值', value: 2, margin: { marginTop: 0, marginLeft: 0, marginBottom: 0, marginRight: 0 } },
        { name: '自定义', value: 3, margin: { marginTop: 1, marginLeft: 1, marginBottom: 1, marginRight: 1 } }
    ],
    message: {
        validate: {
            margin: '请选择边距!',
            pageSize: '请选择纸张!',
            distribution: '请选择打印方向!'
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var IDE_CMP$1 = {
    schemaManager: SCHEME_MANAGEMENT$1,
    chartsSchemaManager: CHARTS_SCHEME_MANAGEMENT$1,
    conditionSchemaManager: CONDITION_SCHEMA_MANAGER$1,
    paramsMapping: PARAMS_MAPPING_OPTION$1,
    drillDown: DRILL_DOWN_OPTION$1,
    printOption: PRINT_INTEGRATION_ZH_CHS$1
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var QUERY_FRAMEWORK$1 = {
    exportName: '导出',
    btnExportText: 'Excel(文本)',
    btnExportValue: 'Excel(值)',
    pageInfo: {
        totalPreInfo: '共',
        totalEndInfo: '条'
    },
    printName: '打印',
    refreshName: '刷新',
    filterName: '筛选',
    btnManager: '管理',
    dimensionSetting: '维度设置',
    btnCondition: '查询条件',
    btnQuery: '查询',
    btnSaveAsSchema: '另存为方案',
    btnSaveSchema: '保存方案',
    btnPrintForm: '单据打印',
    conditionManagerListColumn: ['方案范围', '默认', '操作'],
    btnOk: '确定',
    btnCancel: '取消',
    dialogTitle: '查询条件',
    dialogTitle1: '方案管理',
    dialogTitle2: '另存为方案',
    tabName1: '常用条件',
    tabName2: '高级查询',
    schemaName: '方案名称',
    ispreName: '系统预制',
    defaultSchemaName: '设置为默认',
    schemaType: {
        public: '公共',
        personal: '个人'
    },
    defaultName: '默认',
    message: {
        saveas: '方案名称已存在, 请修改!',
        lengthValidate: '方案名称最多不超过15个字符!',
        save: {
            success: '保存成功!',
            error: '保存失败!',
        },
        update: {
            success: '更新成功!',
            error: '更新失败!',
        },
        delete: {
            success: '删除成功!',
            error: '删除失败!',
        },
        info: {
            title: '提示信息',
            delete: '请选择要删除的方案!'
        },
        operate: {
            success: '操作成功!',
            error: '操作失敗!',
        },
        validate: {
            judgeUpdate: '格式方案已修改, 是否放弃修改?',
            notNull: '方案名称不可以为空!',
            exists: '方案名称已存在!',
            beMetadataIdEmpty: '单据打印对应的BE元数据ID为空!',
            formIdFieldEmpty: '单据打印表单主键字段为空!',
            unselected: '请选择要打印的数据!'
        }
    },
    clearCache: '清空缓存',
    btnExpand: '全部展开',
    btnCollapse: '全部折叠',
    btnClearSort: '清除排序'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var SPREAD$1 = {
    crossTable: {
        dimensionManager: '维度设置',
        dimensionName: [
            '未选择', '行维度', '列维度', '值维度'
        ],
        columnManager: '列设置',
        columnSetting: [
            '未选择', '列设置', '列分组'
        ],
        valueDimensionTitle: '请将数值字段拖到这里',
        rowDimensionTitle: '请将文本字段拖到这里',
        colDimensionTitle: '请将文本字段拖到这里',
    },
    aggreateText: '汇总',
    message: {
        export: '系统不支持预制方案的导出，请将预制方案保存后再执行导出操作!',
        print: '系统不支持预制方案的打印，请将预制方案保存后再执行打印操作!',
        preview: '系统不支持预制方案的打印预览，请将预制方案保存后再执行打印预览操作!',
        aysnExport: '数据量比较大，系统进行异步导出，导出完成后请到系统消息中进行查看!',
        clearCache: '清除缓存成功!',
        extend: {
            query: {
                before: '查询前',
                after: '查询后',
            },
            dataLoading: {
                before: '数据加载前',
                after: '数据加载后'
            },
            render: {
                before: '渲染前',
                after: '渲染后'
            }
        },
        click: {
            select: '行选中事件',
            cellSelect: '单元格选中事件',
        },
        validate: {
            schemaInfo: '请维护当前查询的格式方案!',
            rowDimension: '请维护当前查询对应格式方案的行维度!',
            valueDimension: '请维护当前查询对应格式方案的值维度',
            nodata: '当前查询没有数据',
            sort: '无法对系统预制格式进行修改, 请先另存方案!',
            qdpQueryExport: '没有导出操作权限！',
            qdpQueryPrint: '没有打印操作权限！',
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var Formular$1 = {
    define: {
        labelSaveFormular: '保存公式',
        labelSaveAsFormular: '另存为公式',
        labelDeleteFormular: '删除公式',
        labelGeneraterFormular: '形成公式',
        labelQueryFunc: '业务查询函数',
        labelBasicInfo: '基本信息',
        labelFormularName: '公式名称',
        labelIsGroup: '是否分组',
        labelParamInfo: '参数信息',
        labelResultColInfo: '结果列信息',
        resultSortColumns: [
            {
                id: '0',
                label: '无'
            },
            {
                id: '1',
                label: '升序'
            },
            {
                id: '2',
                label: '降序'
            }
        ],
        resultColVisibleColumns: [
            {
                id: '0',
                label: '否'
            },
            {
                id: '1',
                label: '是'
            }
        ],
        columntitles: {
            'resultcolvisible': '是否取值',
            'resultcode': '取值列',
            'resultdescription': '列描述',
            'resultsort': '取值列排序'
        },
        messageAlert: {
            'nameVerify': '公式名称必填！',
            'funcVerify': '业务查询函数必填！',
            'saveFormular': '公式保存成功！',
            'saveAsFormular': '公式另存为成功！',
            'saveAsFormularFailture': '公式另存为失败！',
            'pleaseSelectQueryFunc': '请选择业务查询函数！',
            'pleaseSelectFormular': '请选择业务查询公式！',
            'resultVerify': '结果列信息中请保留一列作为返回值！',
            'formularResultVerify': '结果列信息中请至少保留一列作为返回值！',
            'combineVerify': '请维护组合公式！',
            'formularIntegrityVerify': '公式不正确，请检查后点击完成!',
            'saveFormularFailure': '保存公式失败！',
            'formularSaveVerify': '请保存公式后点击完成！',
            'deleteFormular': '公式删除成功！',
            'deleteFormularFailure': '公式删除失败！',
            'formularIsProtected': '此公式受保护，无法修改。',
            'formularIsProtectedNotDelete': '此公式受保护，无法删除。'
        }
    },
    defineWizard: {
        labelSelectTemplate: '向导选择',
        labelFormularDefine: '公式定义',
    },
    step: {
        labelPre: '上一步',
        labelNext: '下一步',
        labelFinish: '完成',
    },
    formularSelect: {
        selectionFormular: '[{"value":"0","name":"公式选择"},{"value":"1","name":"公式创建"},{"value":"2","name":"组合公式"}]',
        lableQdpFormular: '业务查询公式',
        messageAlert: {
            'formularSelectVerify': '请选择取数公式！',
            'formularDefineVerify': '请定义取数公式或组合公式！',
        }
    },
    service: {
        lableFormulaWizard: '公式向导',
        messageAlert: {
            'formulaVerify': '当前没有形成有效的公式，确认关闭'
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var QDP_ZH_CHS = {
    ideCmp: IDE_CMP$1,
    queryFramework: QUERY_FRAMEWORK$1,
    spread: SPREAD$1,
    formular: Formular$1
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var TITLE_OPTION$2 = {
    tab: '標題',
    title: '標題',
    subTitleColumnName: [
        '類型', '內容', '級次', '對齊'
    ],
    subTitleData: [
        { label: '副標題', value: 0 },
        { label: '表尾', value: 1 }
    ],
    subTitleAlign: [
        { label: '左', value: 0 },
        { label: '中', value: 1 },
        { label: '右', value: 2 }
    ]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var COLUMN_OPTION$2 = {
    tab: '列',
    btnAdd: '新增',
    btnDelete: '刪除',
    btnUp: '上移',
    btnDown: '下移',
    btnLeft: '左移',
    btnRight: '右移',
    selection1: '基礎信息',
    selection2: '高級信息',
    name: '名稱',
    binding: '綁定字段',
    showType: '顯示類型',
    colWidth: '列寬',
    align: '對齊方式',
    alignData: [
        { name: '左', value: 0 },
        { name: '中', value: 1 },
        { name: '右', value: 2 }
    ],
    show: '顯示',
    fixColumn: '固定列',
    wordWrap: '折行',
    autoFitColumn: '自適應',
    dataShow: '數據展示',
    formatType: '格式化類型',
    formatTypeData: [
        { name: '請選擇', value: 0 },
        { name: '數字', value: 1 },
        { name: '日期', value: 2 },
        { name: '邏輯', value: 3 }
    ],
    dateFormat: '日期格式',
    numberFormat: '數字格式',
    numberField: '精度描述字段',
    enumFormat: '格式串',
    sort: '排序',
    sortData: [
        { name: '請選擇', value: 0 },
        { name: '正序', value: 1 },
        { name: '倒序', value: 2 }
    ],
    group: '是否分組',
    pageTotal: '本頁小計文本',
    pageTotalType: '本頁小計類型',
    groupTotal: '分組小計文本',
    groupSubTotalType: '分組小計類型',
    groupTotalType: '分組合計類型',
    aggregateTotal: '匯總合計文本',
    totalType: '匯總合計類型',
    aggregateData: [
        { name: '請選擇', value: 0 },
        { name: '合計', value: 1 },
        { name: '最小值', value: 2 },
        { name: '最大值', value: 3 },
        { name: '計數', value: 4 },
        { name: '平均值', value: 5 }
    ],
    dimension: '交叉維度',
    dimensionData: [
        { name: '請選擇', value: 0 },
        { name: '行維度', value: 1 },
        { name: '列緯度', value: 2 },
        { name: '值緯度', value: 3 }
    ],
    enumType: [
        { name: '索引', value: '0' },
        { name: '編號', value: '1' },
        { name: '名稱', value: '2' }
    ],
    groupColor: '分組行顏色',
    dataShowOption: [
        { name: '默認', value: -1 },
        { name: '自適應', value: 1 },
        { name: '折行', value: 2 }
    ],
    isRed: '負數紅字'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var STYLE_OPTION$2 = {
    tab: '樣式',
    column: '欄目',
    styleTitle: '標題',
    styleSubTitle: '副標題',
    styleHeader: '表頭',
    styleBody: '表體',
    styleFooter: '表尾',
    rowHeight: '行高',
    textStyle: '文本樣式',
    preview: '預覽',
    fontFamilyData: [
        { name: '宋體', value: 0 },
        { name: '新宋體', value: 1 },
        { name: '仿宋', value: 2 },
        { name: '楷體', value: 3 },
        { name: '黑體', value: 4 },
        { name: 'ARIAL', value: 5 },
        { name: 'ROMAN', value: 6 },
        { name: 'TIMES NEW ROMAN', value: 7 },
        { name: 'VERDANA', value: 8 },
        { name: 'Microsoft YaHei', value: 9 }
    ]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var OTHER_OPTION$2 = {
    tab: '其它',
    selection: '基本信息',
    dataCache: '數據緩存',
    firstLoad: '首次查詢',
    filterRow: '開啟過濾行',
    showHeaderSort: '開啟列頭排序',
    mergeCell: '是否合併單元格',
    pageType: '分頁類型',
    pageSize: '分頁大小',
    delay: '數據延遲加載',
    timing: '數據定時加載',
    export: '導出範圍',
    groupOption: '分支設置',
    rowSubtotal: '行小計',
    colSubtotal: '列小計',
    rowTotal: '行合計',
    colTotal: '列合計',
    showZero: '0值是否顯示',
    expand: '默認展開',
    pageTypeData: [
        { name: '全部加載', value: 0 },
        { name: '分頁加載', value: 1 },
    ],
    rangeData: [
        { name: '全部', value: 0 },
        { name: '當前頁', value: 1 },
        { name: '自定義', value: 2 }
    ],
    treeInfo: '樹表設置',
    treeInfoType: '構建類型',
    treeInfoData: [
        { name: '請選擇', value: 2 },
        { name: '分級碼', value: 0 },
        { name: '自關聯', value: 1 }
    ],
    pathField: '分級碼字段',
    pathStep: '分級碼長度',
    idField: '主鍵字段',
    parentField: '父級字段',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var SCHEME_MANAGEMENT$2 = {
    dailogName: '格式方案',
    formatSchemaManagerName: '格式方案管理',
    formatManager: '格式管理',
    preFormat: '系統預製格式',
    btnCancel: '取消',
    btnOK: '确定',
    btnSave: '保存',
    btnSaveAs: '另存為',
    btnDelete: '刪除',
    btnSaveAsClose: '保存並關閉',
    schemeName: '名稱',
    schemeType: '範圍',
    schemeTypePublic: '公共',
    schemePublic: '公共方案',
    schemeTypePersonal: '個人',
    schemePersonal: '個人方案',
    schemeTypeOrg: '單位',
    schemeOrg: '單位方案',
    schemeDefault: '默認',
    titleOption: TITLE_OPTION$2,
    columnOption: COLUMN_OPTION$2,
    styleOption: STYLE_OPTION$2,
    otherOption: OTHER_OPTION$2,
    message: {
        save: {
            success: '保存成功!',
            error: '保存失敗!',
        },
        update: {
            success: '更新成功!',
            error: '更新失敗!',
        },
        delete: {
            success: '删除成功!',
            error: '删除失敗!',
        },
        info: {
            title: '提示信息',
            delete: '請選擇要刪除的方案!'
        },
        validate: {
            judgeUpdate: '格式方案已修改, 是否放棄修改?',
            notNull: '方案名稱不可以為空!',
            exists: '方案名稱已存在!',
            tooLong: '方案名稱不可以超過100個字符!',
            isProtected: '系統預製方案無法修改和刪除!',
            crosstabValueDimension: '值維度必須是數值列！',
            qdpModifyPublicSchema: '沒有操作公共方案的權限！',
            textColumnSort: '列是大文本，不支持排序操作！'
        }
    },
    basicInfoName: '基本信息'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CHARTS_TYPE_OPTION$2 = {
    tab: '類型',
    chartsType: '圖表類型',
    chartsTypeData: [
        { name: '請選擇', value: -1 },
        { name: '折線圖', value: 'line' },
        { name: '柱狀圖', value: 'bar' },
        { name: '餅圖', value: 'pie' },
        { name: '面積圖', value: 'line1' },
        { name: '散點圖', value: 'scatter' },
        { name: '地圖散點', value: 'effectScatter' }
    ]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CHARTS_DATA_OPTION$2 = {
    tab: '數據',
    xAxis: 'x軸',
    xPositionData: [
        { name: '下', value: 'bottom' },
        { name: '上', value: 'top' }
    ],
    yAxis: 'y軸',
    yPositionData: [
        { name: '左', value: 'left' },
        { name: '右', value: 'right' }
    ],
    axisTypeData: [
        { name: '分類', value: 'category' },
        { name: '值', value: 'value' }
    ],
    type: '類型',
    position: '位置',
    binding: '綁定字段',
    category: '分類',
    numerical1: '數值1',
    numerical2: '數值2',
    // 经度
    longitude: '經度',
    // 纬度
    latitude: '緯度',
    barTypeData: [
        { name: '單列', value: 1 },
        { name: '多列', value: 2 },
        { name: '疊加', value: 3 },
    ],
    name: '名称',
    select: '請選擇',
    dimensionSetting: '緯度設置',
    dimension: '緯度',
    indicator: '指標'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CHARTS_STYLE_OPTION$2 = {
    tab: '樣式',
    titleOption: '標題設置',
    title: '標題',
    select: '請選擇',
    xAxisOption: 'x軸設置',
    xAxisPosition: 'x軸位置',
    yAxisPosition: 'y軸位置',
    xAxisAlignData: [
        { name: '左', value: 'left' },
        { name: '中', value: 'center' },
        { name: '右', value: 'right' }
    ],
    yAxisOption: 'y軸设置',
    yAxisAlignData: [
        { name: '上', value: 'top' },
        { name: '中', value: 'middle' },
        { name: '下', value: 'bottom' }
    ],
    fontFamilyData: [
        { name: 'Microsoft YaHei', value: 'Microsoft YaHei' },
        { name: 'sans-serif', value: 'sans-serif' },
        { name: 'serif', value: 'serif' },
        { name: 'monospace', value: 'monospace' },
        { name: 'Arial', value: 'Arial' },
        { name: 'Courier New', value: 'Courier New' }
    ],
    rowHeight: '行高',
    contentStyle: '文本樣式',
    multiDataOption: '多維數據設置',
    dataMultiple: [
        { name: '並排', value: 'side' },
        { name: '堆疊', value: 'pile' }
    ],
    multiDataType: '類型',
    legend: '圖例',
    legendAlgin: '對齊方式',
    legendAlignData: [
        { name: '自適應', value: 'auto' },
        { name: '左', value: 'left' },
        { name: '右', value: 'right' }
    ],
    legendOrient: '佈局方向',
    orientData: [
        { name: '水平', value: 'horizontal' },
        { name: '垂直', value: 'vertical' }
    ],
    legendGap: '間距',
    itemWidth: '寬度',
    itemHeight: '高度',
    visualMap: '視覺地圖',
    visualMapType: '類型',
    visualMapTypeData: [
        { name: 'continuous', value: 'continuous' },
        { name: 'piecewise', value: 'piecewise' }
    ],
    visualMapOrient: '佈局方向',
    visualMapPosition: '位置',
    visualMapTop: 'Top',
    visualMapBottom: 'Bottom',
    visualMapLeft: 'Left',
    visualMapRight: 'right',
    continuous: 'Continuous',
    piecewise: 'Piecewise',
    otherStyle: '其他',
    customStyle: '自定義',
    validate: {
        echartsCustomOptions: '非法的Json對象格式!'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CHARTS_SCHEME_MANAGEMENT$2 = {
    chartsSchemaManagerName: '格式方案管理',
    chartsType: CHARTS_TYPE_OPTION$2,
    chartsData: CHARTS_DATA_OPTION$2,
    chartsStyle: CHARTS_STYLE_OPTION$2
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var PARAMS_MAPPING_OPTION$2 = {
    paramOption: '參數設置',
    btnAdd: '新增',
    btnDelete: '删除',
    btnCancel: '取消',
    btnOK: '確定',
    columnHeader: [
        '參數名', '映射類別', '源參數', '常量', '表達式', '作為查詢參數'
    ],
    mappingTypeData: [
        { name: '字段', value: 0 },
        { name: '常量', value: 1 },
        { name: '表達式', value: 2 }
    ],
    message: {
        validate: {
            select: '請選擇一條要刪除的記錄!'
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var DRILL_DOWN_OPTION$2 = {
    drillDownName: '聯查管理',
    btnAdd: '新增',
    btnDelete: '刪除',
    btnCancel: '取消',
    btnOK: '確定',
    menuBtn: '菜單',
    appBtn: '應用',
    columnHeader: [
        '功能編號', '標題', '聯查主鍵依據', '聯查展現形式', '可見表達式', '雙擊聯查', '超鏈字段'
    ],
    paramColumnHeader: [
        '參數', '名稱', '值', '是否為表達式'
    ],
    displayStyle: [
        { 'name': '右鍵聯查', value: 1 },
        { 'name': '超鏈聯查', value: 2 }
    ],
    enumExpress: [
        '是', '否'
    ]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CONDITION_SCHEMA_MANAGER$2 = {
    conditionSchemeManagerName: '查詢條件方案管理',
    schemeSave: '保存方案',
    schemeManagement: '方案管理',
    btnDelete: '刪除',
    btnSave: '保存',
    schemeName: '方案名称',
    schemeType: '方案類別',
    saveScheme: '保存方案',
    public: '公共',
    personal: '個人',
    default: '默認',
    noDefault: '非默認',
    message: {
        save: {
            success: '保存成功!',
            error: '保存失敗!'
        },
        update: {
            success: '修改默認方案成功!',
            error: '修改失敗!請聯繫管理員!'
        },
        delete: {
            success: '刪除成功!',
            error: '删除失敗!'
        },
        validate: {
            delete: '默認方案不允許刪除!',
            longLimit: '方案名稱過長!',
            repeatName: '方案名稱已存在!',
            select: '請選擇一條要刪除的記錄!'
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var PRINT_INTEGRATION_ZH_CHS$2 = {
    print: '打印',
    dailogName: '打印設置',
    layout: '佈局',
    paperSize: '紙張尺寸',
    margin: '邊距',
    zoom: '縮放',
    pdfDownload: 'PDF下載',
    printPreview: '打印預覽',
    distributionData: [
        { name: '橫向', value: 1 },
        { name: '縱向', value: 0 }
    ],
    pageSizeData: [
        { name: '信紙', value: 1 },
        { name: 'Tabloid', value: 3 },
        { name: '法律專用紙', value: 5 },
        { name: 'Statement', value: 6 },
        { name: 'A3', value: 8 },
        { name: 'A4', value: 9 },
        { name: 'A5', value: 11 },
        { name: 'B4 (JIS)', value: 12 },
        { name: 'B5 (JIS)', value: 13 }
    ],
    marginData: [
        { name: '默認', value: 0, margin: { marginTop: 1, marginLeft: 1, marginBottom: 1, marginRight: 1 } },
        { name: '无', value: 1, margin: { marginTop: 0, marginLeft: 0, marginBottom: 0, marginRight: 0 } },
        { name: '最小值', value: 2, margin: { marginTop: 0, marginLeft: 0, marginBottom: 0, marginRight: 0 } },
        { name: '自定義', value: 3, margin: { marginTop: 1, marginLeft: 1, marginBottom: 1, marginRight: 1 } }
    ],
    message: {
        validate: {
            margin: '請選擇邊距!',
            pageSize: '請選擇紙張!',
            distribution: '請選擇打印方向!'
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var IDE_CMP$2 = {
    schemaManager: SCHEME_MANAGEMENT$2,
    chartsSchemaManager: CHARTS_SCHEME_MANAGEMENT$2,
    conditionSchemaManager: CONDITION_SCHEMA_MANAGER$2,
    paramsMapping: PARAMS_MAPPING_OPTION$2,
    drillDown: DRILL_DOWN_OPTION$2,
    printOption: PRINT_INTEGRATION_ZH_CHS$2
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var QUERY_FRAMEWORK$2 = {
    exportName: '導出',
    btnExportText: 'Excel(文本)',
    btnExportValue: 'Excel(值)',
    pageInfo: {
        totalPreInfo: '共',
        totalEndInfo: '條'
    },
    printName: '打印',
    refreshName: '刷新',
    filterName: '篩選',
    dimensionSetting: '維度設置',
    btnCondition: '查詢條件',
    btnQuery: '查詢',
    btnSaveAsSchema: '另存為方案',
    btnSaveSchema: '保存方案',
    btnPrintForm: '單據打印',
    conditionManagerListColumn: ['方案範圍', '默認', '操作'],
    btnOk: '確定',
    btnCancel: '取消',
    dialogTitle: '查詢條件',
    dialogTitle1: '方案管理',
    dialogTitle2: '另存為方案',
    tabName1: '常用條件',
    tabName2: '高級查詢',
    schemaName: '方案名稱',
    ispreName: '系統預製',
    defaultSchemaName: '設置為默認',
    schemaType: {
        public: '公共',
        personal: '個人'
    },
    defaultName: '默認',
    message: {
        saveas: '方案名稱已存在, 請修改!',
        lengthValidate: '方案名稱最多不超過15个字符!',
        save: {
            success: '保存成功!',
            error: '保存失敗!',
        },
        update: {
            success: '更新成功!',
            error: '更新失敗!',
        },
        delete: {
            success: '删除成功!',
            error: '删除失敗!',
        },
        info: {
            title: '提示信息',
            delete: '請選擇要刪除的方案!'
        },
        operate: {
            success: '操作成功!',
            error: '操作失敗!',
        },
        validate: {
            judgeUpdate: '格式方案已修改, 是否放棄修改?',
            notNull: '方案名稱不可以為空!',
            exists: '方案名稱已存在!',
            beMetadataIdEmpty: '單據打印對應的BE元數據ID為空!',
            formIdFieldEmpty: '單據打印表單主鍵字段為空!',
            unselected: '請選擇要打印的數據!'
        }
    },
    clearCache: '清空緩存',
    btnExpand: '全部展開',
    btnCollapse: '全部折疊',
    btnClearSort: '清除排序'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var SPREAD$2 = {
    crossTable: {
        dimensionManager: '維度設置',
        dimensionName: [
            '未選擇', '行維度', '列維度', '值維度'
        ],
        columnManager: '列設置',
        columnSetting: [
            '未選擇', '列設置', '列分組'
        ],
        valueDimensionTitle: '請將數值字段拖到這裡',
        rowDimensionTitle: '請將文本字段拖到這裡',
        colDimensionTitle: '請將文本字段拖到這裡',
    },
    aggreateText: '匯總',
    message: {
        export: '系統預製方案不支持導出，請將預製方案另存後再執行導出操作!',
        print: '系統預製方案不支持打印，請將預製方案另存後再執行打印操作!',
        preview: '系統預製方案不支持打印預覽，請將預製方案另存後再執行打印預覽操作!',
        aysnExport: '數據量比較大，系統進行異步導出，導出完成後請到系統消息中進行查看！',
        clearCache: '清除緩存成功!',
        extend: {
            query: {
                before: '查詢前',
                after: '查詢后',
            },
            dataLoading: {
                before: '數據加載前',
                after: '數據加載后'
            },
            render: {
                before: '渲染前',
                after: '渲染后'
            }
        },
        click: {
            select: '行選中事件',
            cellSelect: '單元格選中事件',
        },
        validate: {
            schemaInfo: '請維護當前查詢的格式方案!',
            rowDimension: '請維護當前查詢對應格式方案的行維度!',
            valueDimension: '請維護當前查詢對應格式方案的值維度',
            nodata: '當前查詢沒有數據',
            sort: '無法對系統預製格式方案進行修改，請另存方案!',
            qdpQueryExport: '没有導出操作權限！',
            qdpQueryPrint: '没有打印操作權限！',
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var Formular$2 = {
    define: {
        labelSaveFormular: '保存公式',
        labelSaveAsFormular: '另存爲公式',
        labelDeleteFormular: '刪除公式',
        labelGeneraterFormular: '形成公式',
        labelQueryFunc: '業務查詢函數',
        labelBasicInfo: '基本信息',
        labelFormularName: '公式名稱',
        labelIsGroup: '是否分組',
        labelParamInfo: '參數信息',
        labelResultColInfo: '結果列信息',
        resultSortColumns: [
            {
                id: '0',
                label: '無'
            },
            {
                id: '1',
                label: '升序'
            },
            {
                id: '2',
                label: '降序'
            }
        ],
        resultColVisibleColumns: [
            {
                id: '0',
                label: '否'
            },
            {
                id: '1',
                label: '是'
            }
        ],
        columntitles: {
            'resultcolvisible': '是否取值',
            'resultcode': '取值列',
            'resultdescription': '列描述',
            'resultsort': '取值列排序'
        },
        messageAlert: {
            'nameVerify': '公式名稱必填！',
            'funcVerify': '業務查詢函數必填！',
            'saveFormular': '公式保存成功！',
            'saveAsFormular': '公式另存爲保存成功！',
            'saveAsFormularFailture': '公式另存爲失敗！',
            'pleaseSelectQueryFunc': '請選擇業務查詢函數！',
            'pleaseSelectFormular': '請選擇業務查詢公式！',
            'resultVerify': '結果列信息中請保留一列作為返回值！',
            'formularResultVerify': '結果列信息中請至少保留一列作爲返回值！',
            'combineVerify': '請維護組合公式！',
            'formularIntegrityVerify': '公式不正確，請檢查後點擊完成!',
            'saveFormularFailure': '公式保存失敗！',
            'formularSaveVerify': '請保存公式後點擊完成！',
            'deleteFormular': '公式刪除成功！',
            'deleteFormularFailure': '公式刪除失敗！',
            'formularIsProtected': '此公式受保護，無法修改。',
            'formularIsProtectedNotDelete': '此公式受保護，無法刪除。'
        }
    },
    defineWizard: {
        labelSelectTemplate: '嚮導選擇',
        labelFormularDefine: '公式定義',
    },
    step: {
        labelPre: '上一步',
        labelNext: '下一步',
        labelFinish: '完成',
    },
    formularSelect: {
        selectionFormular: '[{"value":"0","name":"公式選擇"},{"value":"1","name":"公式創建"},{"value":"2","name":"組合公式"}]',
        lableQdpFormular: '業務查詢公式',
        messageAlert: {
            'formularSelectVerify': '請選擇取數公式！',
            'formularDefineVerify': '請定義取數公式或組合公式！',
        }
    },
    service: {
        lableFormulaWizard: '公式嚮導',
        messageAlert: {
            'formulaVerify': '當前沒有形成有效的公式，確認關閉'
        }
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var QDP_ZH_CHT = {
    ideCmp: IDE_CMP$2,
    queryFramework: QUERY_FRAMEWORK$2,
    spread: SPREAD$2,
    formular: Formular$2
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var QDP_LOCALIZE = {
    'zh-chs': QDP_ZH_CHS,
    'zh-cn': QDP_ZH_CHS,
    'zh': QDP_ZH_CHS,
    'zh-cht': QDP_ZH_CHT,
    'en-us': QDP_EN_US,
    'en': QDP_EN_US
};
/** @type {?} */
var DEFAULT_LANGUAGE = 'zh-chs';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var LocalizeService = /** @class */ (function () {
    function LocalizeService(localeId, utils) {
        this.localeId = localeId;
        this.utils = utils;
        if (!localeId) {
            localeId = DEFAULT_LANGUAGE;
            this.language = 'zh-CHS';
        }
        else {
            this.language = localeId;
        }
        if (!QDP_LOCALIZE[localeId.toLowerCase()]) {
            localeId = DEFAULT_LANGUAGE;
        }
        this.languageData = QDP_LOCALIZE[localeId.toLowerCase()];
    }
    /**
     * @param {?} propertyName
     * @return {?}
     */
    LocalizeService.prototype.getValue = /**
     * @param {?} propertyName
     * @return {?}
     */
    function (propertyName) {
        /** @type {?} */
        var val = this.utils.getValue(propertyName, this.languageData);
        if (val) {
            return val;
        }
        else {
            console.warn("Can't find language Data for the " + propertyName);
        }
    };
    LocalizeService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    LocalizeService.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: CommonUtils }
    ]; };
    /** @nocollapse */ LocalizeService.ngInjectableDef = defineInjectable({ factory: function LocalizeService_Factory() { return new LocalizeService(inject(LOCALE_ID), inject(CommonUtils)); }, token: LocalizeService, providedIn: "root" });
    return LocalizeService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var LocalizeModule = /** @class */ (function () {
    function LocalizeModule() {
    }
    /**
     * @return {?}
     */
    LocalizeModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: LocalizeModule,
            providers: [
                LocalizeService
            ]
        };
    };
    LocalizeModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [
                        FarrisCommonModule.forRoot()
                    ],
                    exports: []
                },] }
    ];
    return LocalizeModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { QDP_EN_US, QDP_ZH_CHS, LocalizeService, LocalizeModule, QDP_LOCALIZE, DEFAULT_LANGUAGE, Formular as ɵq, CHARTS_DATA_OPTION as ɵi, CHARTS_STYLE_OPTION as ɵj, CHARTS_TYPE_OPTION as ɵh, CHARTS_SCHEME_MANAGEMENT as ɵg, CONDITION_SCHEMA_MANAGER as ɵk, DRILL_DOWN_OPTION as ɵm, IDE_CMP as ɵa, PARAMS_MAPPING_OPTION as ɵl, PRINT_INTEGRATION_ZH_CHS as ɵn, COLUMN_OPTION as ɵd, SCHEME_MANAGEMENT as ɵb, OTHER_OPTION as ɵf, STYLE_OPTION as ɵe, TITLE_OPTION as ɵc, QUERY_FRAMEWORK as ɵo, SPREAD as ɵp, Formular$1 as ɵbh, CHARTS_DATA_OPTION$1 as ɵz, CHARTS_STYLE_OPTION$1 as ɵba, CHARTS_TYPE_OPTION$1 as ɵy, CHARTS_SCHEME_MANAGEMENT$1 as ɵx, CONDITION_SCHEMA_MANAGER$1 as ɵbb, DRILL_DOWN_OPTION$1 as ɵbd, IDE_CMP$1 as ɵr, PARAMS_MAPPING_OPTION$1 as ɵbc, PRINT_INTEGRATION_ZH_CHS$1 as ɵbe, COLUMN_OPTION$1 as ɵu, SCHEME_MANAGEMENT$1 as ɵs, OTHER_OPTION$1 as ɵw, STYLE_OPTION$1 as ɵv, TITLE_OPTION$1 as ɵt, QUERY_FRAMEWORK$1 as ɵbf, SPREAD$1 as ɵbg, Formular$2 as ɵbz, CHARTS_DATA_OPTION$2 as ɵbr, CHARTS_STYLE_OPTION$2 as ɵbs, CHARTS_TYPE_OPTION$2 as ɵbq, CHARTS_SCHEME_MANAGEMENT$2 as ɵbp, CONDITION_SCHEMA_MANAGER$2 as ɵbt, DRILL_DOWN_OPTION$2 as ɵbv, IDE_CMP$2 as ɵbj, PARAMS_MAPPING_OPTION$2 as ɵbu, PRINT_INTEGRATION_ZH_CHS$2 as ɵbw, COLUMN_OPTION$2 as ɵbm, SCHEME_MANAGEMENT$2 as ɵbk, OTHER_OPTION$2 as ɵbo, STYLE_OPTION$2 as ɵbn, TITLE_OPTION$2 as ɵbl, QDP_ZH_CHT as ɵbi, QUERY_FRAMEWORK$2 as ɵbx, SPREAD$2 as ɵby };
