// @farris/ui-datagrid-settings/esm2015/lib/dbl-click.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Output, EventEmitter } from "@angular/core";
import { fromEvent, race } from 'rxjs';
import { buffer, bufferCount, debounceTime, filter, first, repeat } from 'rxjs/operators';
export class FarrisDblClickDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        this.dblClickHandler = new EventEmitter();
        this._click$ = null;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        /** @type {?} */
        const doubleClickDuration = 250;
        /** @type {?} */
        const leftClick$ = fromEvent(this.el.nativeElement, 'click').pipe(filter((/**
         * @param {?} event
         * @return {?}
         */
        (event) => event.button === 0)));
        /** @type {?} */
        const debounce$ = leftClick$.pipe(debounceTime(doubleClickDuration));
        /** @type {?} */
        const clickLimit$ = leftClick$.pipe(bufferCount(2));
        /** @type {?} */
        const bufferGate$ = race(debounce$, clickLimit$)
            .pipe(first(), repeat());
        this._click$ = leftClick$.pipe(buffer(bufferGate$), filter((/**
         * @param {?} clicks
         * @return {?}
         */
        clicks => {
            return clicks.length > 1;
        }))).subscribe((/**
         * @param {?} clicks
         * @return {?}
         */
        clicks => {
            this.dblClickHandler.emit(clicks[0]);
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this._click$) {
            this._click$.unsubscribe();
            this._click$ = null;
        }
    }
}
FarrisDblClickDirective.decorators = [
    { type: Directive, args: [{
                selector: '[f-dbl-click]'
            },] }
];
/** @nocollapse */
FarrisDblClickDirective.ctorParameters = () => [
    { type: ElementRef }
];
FarrisDblClickDirective.propDecorators = {
    options: [{ type: Input, args: ['f-dbl-click',] }],
    dblClickHandler: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FarrisDblClickDirective.prototype.options;
    /** @type {?} */
    FarrisDblClickDirective.prototype.dblClickHandler;
    /**
     * @type {?}
     * @private
     */
    FarrisDblClickDirective.prototype._click$;
    /**
     * @type {?}
     * @private
     */
    FarrisDblClickDirective.prototype.el;
}
