// @farris/mobile-cli/lib/config/libs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.libOutputOptions = exports.libInputOptions = void 0;
const vue_1 = __importDefault(require("./vue"));
const vue_router_1 = __importDefault(require("./vue-router"));
const vant_1 = __importDefault(require("./vant"));
const axios_1 = __importDefault(require("./axios"));
const dayjs_1 = __importDefault(require("./dayjs"));
const quill_1 = __importDefault(require("./quill"));
const vue_compiler_1 = __importDefault(require("./vue-compiler"));
exports.libInputOptions = {
    vue: vue_1.default.input,
    vant: vant_1.default.input,
    'vue-router': vue_router_1.default.input,
    axios: axios_1.default.input,
    dayjs: dayjs_1.default.input,
    quill: quill_1.default.input,
    'vue-compiler': vue_compiler_1.default.input,
};
exports.libOutputOptions = {
    vue: vue_1.default.output,
    vant: vant_1.default.output,
    'vue-router': vue_router_1.default.output,
    axios: axios_1.default.output,
    dayjs: dayjs_1.default.output,
    quill: quill_1.default.output,
    'vue-compiler': vue_compiler_1.default.output,
};
exports.default = { libInputOptions: exports.libInputOptions, libOutputOptions: exports.libOutputOptions };
