// @gsp-lcm/bo-creator/esm5/lib/views/bo-creator/modal/su-modal/su-creator-modal.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/views/bo-creator/modal/su-modal/su-creator-modal.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output, Input, ViewChild } from '@angular/core';
import { BusinessObjectEntity, BusinessObjectDtService } from '@gsp-lcm/bo-dt-service';
import { BoCreatorUtils } from '../../utils/bo-create-utils';
import { NotifyService } from '@farris/ui-notify';
import { ComboListComponent } from '@farris/ui-combo-list';
import { ValidatorResult } from '../../common/validation-result';
var SuCreatorModalComponent = /** @class */ (function () {
    function SuCreatorModalComponent(boCreatorUtils, boDtService, notifyService) {
        this.boCreatorUtils = boCreatorUtils;
        this.boDtService = boDtService;
        this.notifyService = notifyService;
        this.suConfirmEmitter = new EventEmitter();
        /**
         * 当前BO节点
         */
        this.currentBo = new BusinessObjectEntity();
        /**
         * 关键应用集合
         */
        this.items = [];
        /**
         * 选择应用节点下的所有服务单元集合
         */
        this.suList = [];
        /**
         * 所有业务对象节点集合
         */
        this.data = [];
        this.parent = new BusinessObjectEntity();
        this.labelLength = '2';
        this.inputLength = '9';
    }
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initData();
    };
    /**
     * @private
     * @return {?}
     */
    SuCreatorModalComponent.prototype.initData = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.boDtService.getBusinessObjects().subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.data = data;
            _this.items = data.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.layer === 2; }));
            _this.suList = data.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.layer === 3; }));
        }));
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SuCreatorModalComponent.prototype.appSelectChanged = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.parent = $event.data;
    };
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.create = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var bo = this.getCurrentBo();
        if (!bo) {
            return;
        }
        this.boDtService.save(bo).subscribe((/**
         * @return {?}
         */
        function () {
            _this.suConfirmEmitter.emit(bo);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.notifyService.error('创建失败');
            throw error;
        }));
    };
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.getCurrentBo = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = this.suInfoValidator();
        if (result.illegal) {
            this.notifyService.warning(result.message);
            return null;
        }
        this.currentBo.id = this.boCreatorUtils.newGuid();
        this.currentBo.layer = 3;
        this.currentBo.parentID = this.parentId;
        this.currentBo.sysInit = '0';
        this.currentBo.isSysInit = false;
        this.currentBo.isDetail = '0';
        /** @type {?} */
        var max = this.boCreatorUtils.getMaxOrder(this.suList, this.currentBo.parentID);
        this.currentBo.sortOrder = max + 1;
        return Object.assign(new BusinessObjectEntity(), this.currentBo);
    };
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.suInfoValidator = /**
     * @return {?}
     */
    function () {
        //新增信息空校验
        /** @type {?} */
        var result = this.emptyValidator();
        if (result.illegal) {
            return result;
        }
        //编号，名称命名规范性校验
        result = this.namingConventionValidator();
        if (result.illegal) {
            return result;
        }
        //APP编号重复校验
        return this.repeatValidator();
    };
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.emptyValidator = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = new ValidatorResult();
        if (!this.parentId) {
            result.illegal = true;
            result.message = "\u5173\u952E\u5E94\u7528\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        if (!this.currentBo.code) {
            result.illegal = true;
            result.message = "\u670D\u52A1\u5355\u5143\u7684\u7F16\u53F7\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        if (!this.currentBo.name) {
            result.illegal = true;
            result.message = "\u670D\u52A1\u5355\u5143\u7684\u540D\u79F0\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        return result;
    };
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.namingConventionValidator = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = new ValidatorResult();
        /** @type {?} */
        var codeRegx = new RegExp('^[a-zA-Z\x7f-\xff][a-zA-Z0-9\x7f-\xff]*$');
        if (!codeRegx.test(this.currentBo.code)) {
            result.illegal = true;
            result.message = '服务单元的编号只能包含字母,数字,并且不可以用数字开头!';
            return result;
        }
        /** @type {?} */
        var nameRegx = new RegExp('[`~!@#$^&*()=|{}\':;\',\\[\\].<>/?~！@#￥……&*（）——|{}【】‘；：”“\'。，、？]');
        if (nameRegx.test(this.currentBo.name)) {
            result.illegal = true;
            result.message = '服务单元的名称包含非法字符，请重新输入!';
            return result;
        }
        return result;
    };
    /**
     * @return {?}
     */
    SuCreatorModalComponent.prototype.repeatValidator = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var result = new ValidatorResult();
        if (!this.suList || this.suList.length <= 0) {
            return result;
        }
        /** @type {?} */
        var repeatCodeBO = this.suList.find((/**
         * @param {?} bo
         * @return {?}
         */
        function (bo) { return bo.code.toLowerCase() === _this.currentBo.code.toLowerCase() && bo.layer == 3; }));
        if (repeatCodeBO) {
            result.illegal = true;
            result.message = '微服务的编号不能与已有微服务编号重复，编号为：' + this.currentBo.code;
            return result;
        }
        return result;
    };
    SuCreatorModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-su-creator-modal',
                    template: "<div class=\"cloumn\" style=\"margin-top: 28px;\">\r\n  <div class=\"row suInfo\">\r\n    <label class=\"col-form-label ps-col-form-label\">\r\n      <span class=\"farris-label-info text-danger\">*</span>\r\n      <span class=\"farris-label-text\">\u5173\u952E\u5E94\u7528</span>\r\n    </label>\r\n    <div class=\"col-{{inputLength}}\">\r\n      <farris-combo-list [(ngModel)]=\"parentId\" [idField]=\"'id'\" [textField]=\"'name'\" [data]=\"items\"\r\n        (selectChange)=\"appSelectChanged($event)\" [panelHeight]=\"280\" [editable]=\"false\"></farris-combo-list>\r\n    </div>\r\n  </div>\r\n\r\n  <div class=\"row suInfo\">\r\n    <!-- <div class=\"col-{{labelLength}}\">\u7F16\u53F7</div> -->\r\n    <label class=\"col-form-label ps-col-form-label\">\r\n      <span class=\"farris-label-info text-danger\">*</span>\r\n      <span class=\"farris-label-text\">\u7F16\u53F7</span>\r\n    </label>\r\n    <div class=\"col-{{inputLength}}\">\r\n      <input-group [(ngModel)]=\"currentBo.code\" #boCode=\"ngModel\" required></input-group>\r\n    </div>\r\n    <!-- <div [hidden]=\"boCode.valid || boCode.pristine\" class=\"alert-danger\">\r\n      \u7F16\u53F7\u4E0D\u80FD\u4E3A\u7A7A\r\n    </div> -->\r\n  </div>\r\n  <div class=\"row suInfo\">\r\n    <label class=\"col-form-label ps-col-form-label\">\r\n      <span class=\"farris-label-info text-danger\">*</span>\r\n      <span class=\"farris-label-text\">\u540D\u79F0</span>\r\n    </label>\r\n    <!-- <div class=\"col-{{labelLength}}\">\u540D\u79F0</div> -->\r\n    <div class=\"col-{{inputLength}}\">\r\n      <input-group [(ngModel)]=\"currentBo.name\" #boName=\"ngModel\" required></input-group>\r\n    </div>\r\n    <!-- <div [hidden]=\"boName.valid || boName.pristine\" class=\"alert-danger\">\r\n      \u540D\u79F0\u4E0D\u80FD\u4E3A\u7A7A\r\n    </div> -->\r\n  </div>\r\n\r\n</div>",
                    styles: [".sucreator-wrapper{background:rgba(255,255,255,.8);width:95%;margin:40px auto 0;max-width:480px;padding:24px 18px;border-radius:8px;position:relative;z-index:100}.suInfo{margin:0 0 16px 20px}.sucreator-wrapper .col-form-label,.sucreator-wrapper .form-control{padding-top:3px;padding-bottom:3px}.sucreator-wrapper .col-form-label{width:70px;margin-right:12px;text-align:right;flex-shrink:0}.bocreator{background:#fff;box-shadow:0 2px 16px 0 rgba(0,0,0,.1);border-radius:4px}.boCreatorPanel{background-color:#fff}div .ps-col-form-label{width:94px;text-align:right}"]
                }] }
    ];
    /** @nocollapse */
    SuCreatorModalComponent.ctorParameters = function () { return [
        { type: BoCreatorUtils },
        { type: BusinessObjectDtService },
        { type: NotifyService }
    ]; };
    SuCreatorModalComponent.propDecorators = {
        parentId: [{ type: Input }],
        suConfirmEmitter: [{ type: Output }],
        appSelector: [{ type: ViewChild, args: ["appSelector",] }]
    };
    return SuCreatorModalComponent;
}());
export { SuCreatorModalComponent };
if (false) {
    /** @type {?} */
    SuCreatorModalComponent.prototype.parentId;
    /** @type {?} */
    SuCreatorModalComponent.prototype.suConfirmEmitter;
    /** @type {?} */
    SuCreatorModalComponent.prototype.appSelector;
    /**
     * 当前BO节点
     * @type {?}
     */
    SuCreatorModalComponent.prototype.currentBo;
    /**
     * 关键应用集合
     * @type {?}
     */
    SuCreatorModalComponent.prototype.items;
    /**
     * 选择应用节点下的所有服务单元集合
     * @type {?}
     */
    SuCreatorModalComponent.prototype.suList;
    /**
     * 所有业务对象节点集合
     * @type {?}
     */
    SuCreatorModalComponent.prototype.data;
    /** @type {?} */
    SuCreatorModalComponent.prototype.parent;
    /** @type {?} */
    SuCreatorModalComponent.prototype.labelLength;
    /** @type {?} */
    SuCreatorModalComponent.prototype.inputLength;
    /**
     * @type {?}
     * @private
     */
    SuCreatorModalComponent.prototype.boCreatorUtils;
    /**
     * @type {?}
     * @private
     */
    SuCreatorModalComponent.prototype.boDtService;
    /**
     * @type {?}
     * @private
     */
    SuCreatorModalComponent.prototype.notifyService;
}
