// @gsp-cmp/ccmp-ui/esm2015/lib/params/component-info.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ComponentConvertor } from '../convertor/convertor';
export class ComponentInfo {
    /**
     * @param {?} obj
     * @return {?}
     */
    init(obj) {
        this.id = obj['id'];
        this.obselete = obj['obselete'];
        this.code = obj['code'];
        this.name = obj['name'];
        this.nameSpace = obj['nameSpace'];
        this.fileName = obj['fileName'];
        this.type = obj['type'];
        this.bizobjectID = obj['bizobjectID'];
        this.language = obj['language'];
        if (obj['metadata']) {
            /** @type {?} */
            let convertor = new ComponentConvertor();
            this.metadata = convertor.initFromObject(obj['metadata']);
        }
    }
}
if (false) {
    /** @type {?} */
    ComponentInfo.prototype.obselete;
    /** @type {?} */
    ComponentInfo.prototype.id;
    /** @type {?} */
    ComponentInfo.prototype.code;
    /** @type {?} */
    ComponentInfo.prototype.name;
    /** @type {?} */
    ComponentInfo.prototype.nameSpace;
    /** @type {?} */
    ComponentInfo.prototype.fileName;
    /** @type {?} */
    ComponentInfo.prototype.type;
    /** @type {?} */
    ComponentInfo.prototype.bizobjectID;
    /** @type {?} */
    ComponentInfo.prototype.language;
    /** @type {?} */
    ComponentInfo.prototype.metadata;
}
