// @farris/ui-filter-panel/esm5/lib/components/radiocomponent/radio.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output } from '@angular/core';
var PanelRadioComponent = /** @class */ (function () {
    function PanelRadioComponent() {
        this.searchHandle = new EventEmitter();
        /**
         * 判断是否是在扩展区域展示
         */
        this.extend = false;
        /**
         * 判断是否显示label
         */
        this.showlabel = true;
        this.selectCheck = [];
        this.radioDataSource = new Array();
    }
    Object.defineProperty(PanelRadioComponent.prototype, "item", {
        get: /**
         * @return {?}
         */
        function () {
            return this._item;
        },
        set: /**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            this._item = data;
            this.getDataSource();
            if (!this._item.hasOwnProperty('value')) {
                this._item.value = { value: '' };
            }
            //this.radioValue = this._item.value.getOriginalValue();
            this.radioValue = this._item.value.value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PanelRadioComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.getDataSource();
    };
    /**
     * @return {?}
     */
    PanelRadioComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.getDataSource();
    };
    /**
     * @return {?}
     */
    PanelRadioComponent.prototype.getDataSource = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this._item && this._item.control) {
            /** @type {?} */
            var radioboxList = (this._item.control);
            this.radioDataSource = Array();
            if (radioboxList.enumValues && radioboxList.enumValues.length > 0) {
                radioboxList.enumValues.forEach((/**
                 * @param {?} element
                 * @return {?}
                 */
                function (element) {
                    _this.radioDataSource.push(Object.assign({}, element));
                }));
            }
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PanelRadioComponent.prototype.onSearch = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event) {
            this._item.value.value = event;
        }
        else {
            this._item.value.value = '';
        }
        // this._item.valueText = this._item.value.value;
        this.getValueText();
        //this.radioValue = this._item.value.getOriginalValue();
        this.radioValue = this._item.value.value;
        this.searchHandle.emit();
    };
    /**
     * @return {?}
     */
    PanelRadioComponent.prototype.getValueText = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var enumValue = (this._item.control).enumValues;
        /** @type {?} */
        var value = (this._item.value).value;
        if (enumValue && enumValue.length) {
            /** @type {?} */
            var texts = enumValue.find((/**
             * @param {?} val
             * @return {?}
             */
            function (val) {
                return val['value'] == value;
            }));
            if (texts) {
                this._item.valueText = texts['name'];
            }
        }
    };
    PanelRadioComponent.decorators = [
        { type: Component, args: [{
                    selector: 'panel-filter-radio',
                    template: "<!-- <farris-radiogroup \r\n[value]=\"radioValue\" \r\n[horizontal]=\"false\" \r\n[data]=\"radioDataSource\"\r\n(changeValue)=\"onSearch($event)\">\r\n</farris-radiogroup> -->\r\n\r\n<div class=\"f-filter-panel-radiogroup\">\r\n    <div class=\"panel-radiogroup-item\" \r\n    [ngClass]=\"{'panel-radiogroup-item-active': radioValue == item.value}\"\r\n    *ngFor=\"let item of radioDataSource\"\r\n    (click)=\"onSearch(item.value)\"\r\n    >\r\n        <span class=\"panel-radiogroup-item-text\">{{item.name}}</span>\r\n    </div>\r\n</div>"
                    // encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    PanelRadioComponent.ctorParameters = function () { return []; };
    PanelRadioComponent.propDecorators = {
        item: [{ type: Input }],
        searchHandle: [{ type: Output, args: ['searchHandle',] }]
    };
    return PanelRadioComponent;
}());
export { PanelRadioComponent };
if (false) {
    /** @type {?} */
    PanelRadioComponent.prototype._item;
    /** @type {?} */
    PanelRadioComponent.prototype.searchHandle;
    /**
     * 判断是否是在扩展区域展示
     * @type {?}
     */
    PanelRadioComponent.prototype.extend;
    /**
     * 判断是否显示label
     * @type {?}
     */
    PanelRadioComponent.prototype.showlabel;
    /** @type {?} */
    PanelRadioComponent.prototype.radioValue;
    /** @type {?} */
    PanelRadioComponent.prototype.selectCheck;
    /** @type {?} */
    PanelRadioComponent.prototype.radioDataSource;
}
