// @edp-bif/runtime-api/esm2015/lib/ref/mxgraph/Handler.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ref/mxgraph/Handler.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { mxgraph } from './mxgraphFactory';
/**
 * Handler
 * @type {?}
 */
const mxCellHighlightType = mxgraph.mxCellHighlight;
export { mxCellHighlightType as mxCellHighlight };
/** @type {?} */
const mxCellMarkerType = mxgraph.mxCellMarker;
export { mxCellMarkerType as mxCellMarker };
/** @type {?} */
const mxCellTrackerType = mxgraph.mxCellTracker;
export { mxCellTrackerType as mxCellTracker };
/** @type {?} */
const mxConnectionHandlerType = mxgraph.mxConnectionHandler;
export { mxConnectionHandlerType as mxConnectionHandler };
/** @type {?} */
const mxConstraintHandlerType = mxgraph.mxConstraintHandler;
export { mxConstraintHandlerType as mxConstraintHandler };
/** @type {?} */
const mxEdgeHandlerType = mxgraph.mxEdgeHandler;
export { mxEdgeHandlerType as mxEdgeHandler };
/** @type {?} */
const mxEdgeSegmentHandlerType = mxgraph.mxEdgeSegmentHandler;
export { mxEdgeSegmentHandlerType as mxEdgeSegmentHandler };
/** @type {?} */
const mxElbowEdgeHandlerType = mxgraph.mxElbowEdgeHandler;
export { mxElbowEdgeHandlerType as mxElbowEdgeHandler };
/** @type {?} */
const mxGraphHandlerType = mxgraph.mxGraphHandler;
export { mxGraphHandlerType as mxGraphHandler };
/** @type {?} */
const mxHandleType = mxgraph.mxHandle;
export { mxHandleType as mxHandle };
/** @type {?} */
const mxKeyHandlerType = mxgraph.mxKeyHandler;
export { mxKeyHandlerType as mxKeyHandler };
/** @type {?} */
const mxPanningHandlerType = mxgraph.mxPanningHandler;
export { mxPanningHandlerType as mxPanningHandler };
/** @type {?} */
const mxPopupMenuHandlerType = mxgraph.mxPopupMenuHandler;
export { mxPopupMenuHandlerType as mxPopupMenuHandler };
/** @type {?} */
const mxRubberbandType = mxgraph.mxRubberband;
export { mxRubberbandType as mxRubberband };
/** @type {?} */
const mxSelectionCellsHandlerType = mxgraph.mxSelectionCellsHandler;
export { mxSelectionCellsHandlerType as mxSelectionCellsHandler };
/** @type {?} */
const mxTooltipHandlerType = mxgraph.mxTooltipHandler;
export { mxTooltipHandlerType as mxTooltipHandler };
/** @type {?} */
const mxVertexHandlerType = mxgraph.mxVertexHandler;
export { mxVertexHandlerType as mxVertexHandler };
