// @edp-bif/runtime-api/esm2015/lib/ref/mxgraph/IO.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ref/mxgraph/IO.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { mxgraph } from './mxgraphFactory';
/**
 * IO
 * @type {?}
 */
const mxCellCodecType = mxgraph.mxCellCodec;
export { mxCellCodecType as mxCellCodec };
/** @type {?} */
const mxChildChangeCodecType = mxgraph.mxChildChangeCodec;
export { mxChildChangeCodecType as mxChildChangeCodec };
/** @type {?} */
const mxDefaultKeyHandlerCodecType = mxgraph.mxDefaultKeyHandlerCodec;
export { mxDefaultKeyHandlerCodecType as mxDefaultKeyHandlerCodec };
/** @type {?} */
const mxDefaultPopupMenuCodecType = mxgraph.mxDefaultPopupMenuCodec;
export { mxDefaultPopupMenuCodecType as mxDefaultPopupMenuCodec };
/** @type {?} */
const mxDefaultToolbarCodecType = mxgraph.mxDefaultToolbarCodec;
export { mxDefaultToolbarCodecType as mxDefaultToolbarCodec };
/** @type {?} */
const mxEditorCodecType = mxgraph.mxEditorCodec;
export { mxEditorCodecType as mxEditorCodec };
/** @type {?} */
const mxGenericChangeCodecType = mxgraph.mxGenericChangeCodec;
export { mxGenericChangeCodecType as mxGenericChangeCodec };
/** @type {?} */
const mxGraphCodecType = mxgraph.mxGraphCodec;
export { mxGraphCodecType as mxGraphCodec };
/** @type {?} */
const mxGraphViewCodecType = mxgraph.mxGraphViewCodec;
export { mxGraphViewCodecType as mxGraphViewCodec };
/** @type {?} */
const mxModelCodecType = mxgraph.mxModelCodec;
export { mxModelCodecType as mxModelCodec };
/** @type {?} */
const mxObjectCodecType = mxgraph.mxObjectCodec;
export { mxObjectCodecType as mxObjectCodec };
/** @type {?} */
const mxRootChangeCodecType = mxgraph.mxRootChangeCodec;
export { mxRootChangeCodecType as mxRootChangeCodec };
/** @type {?} */
const mxStylesheetCodecType = mxgraph.mxStylesheetCodec;
export { mxStylesheetCodecType as mxStylesheetCodec };
/** @type {?} */
const mxTerminalChangeCodecType = mxgraph.mxTerminalChangeCodec;
export { mxTerminalChangeCodecType as mxTerminalChangeCodec };
