// @edp-bif/runtime-api/esm2015/lib/ref/mxgraph/Layout.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ref/mxgraph/Layout.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { mxgraph } from './mxgraphFactory';
/**
 * Layout
 * @type {?}
 */
const mxCircleLayoutType = mxgraph.mxCircleLayout;
export { mxCircleLayoutType as mxCircleLayout };
/** @type {?} */
const mxCompactTreeLayoutType = mxgraph.mxCompactTreeLayout;
export { mxCompactTreeLayoutType as mxCompactTreeLayout };
/** @type {?} */
const mxCompositeLayoutType = mxgraph.mxCompositeLayout;
export { mxCompositeLayoutType as mxCompositeLayout };
/** @type {?} */
const mxEdgeLabelLayoutType = mxgraph.mxEdgeLabelLayout;
export { mxEdgeLabelLayoutType as mxEdgeLabelLayout };
/** @type {?} */
const mxFastOrganicLayoutType = mxgraph.mxFastOrganicLayout;
export { mxFastOrganicLayoutType as mxFastOrganicLayout };
/** @type {?} */
const mxGraphLayoutType = mxgraph.mxGraphLayout;
export { mxGraphLayoutType as mxGraphLayout };
/** @type {?} */
const mxParallelEdgeLayoutType = mxgraph.mxParallelEdgeLayout;
export { mxParallelEdgeLayoutType as mxParallelEdgeLayout };
/** @type {?} */
const mxPartitionLayoutType = mxgraph.mxPartitionLayout;
export { mxPartitionLayoutType as mxPartitionLayout };
/** @type {?} */
const mxRadialTreeLayoutType = mxgraph.mxRadialTreeLayout;
export { mxRadialTreeLayoutType as mxRadialTreeLayout };
/** @type {?} */
const mxStackLayoutType = mxgraph.mxStackLayout;
export { mxStackLayoutType as mxStackLayout };
/**
 * Layout Hierarchical
 * @type {?}
 */
const mxHierarchicalLayoutType = mxgraph.mxHierarchicalLayout;
export { mxHierarchicalLayoutType as mxHierarchicalLayout };
/** @type {?} */
const mxSwimlaneLayoutType = mxgraph.mxSwimlaneLayout;
export { mxSwimlaneLayoutType as mxSwimlaneLayout };
/**
 * Layout Hierarchical Model
 * @type {?}
 */
const mxGraphAbstractHierarchyCellType = mxgraph.mxGraphAbstractHierarchyCell;
export { mxGraphAbstractHierarchyCellType as mxGraphAbstractHierarchyCell };
/** @type {?} */
const mxGraphHierarchyEdgeType = mxgraph.mxGraphHierarchyEdge;
export { mxGraphHierarchyEdgeType as mxGraphHierarchyEdge };
/** @type {?} */
const mxGraphHierarchyModelType = mxgraph.mxGraphHierarchyModel;
export { mxGraphHierarchyModelType as mxGraphHierarchyModel };
/** @type {?} */
const mxGraphHierarchyNodeType = mxgraph.mxGraphHierarchyNode;
export { mxGraphHierarchyNodeType as mxGraphHierarchyNode };
/** @type {?} */
const mxSwimlaneModelType = mxgraph.mxSwimlaneModel;
export { mxSwimlaneModelType as mxSwimlaneModel };
/**
 * Layout Hierarchical Stage
 * @type {?}
 */
const mxCoordinateAssignmentType = mxgraph.mxCoordinateAssignment;
export { mxCoordinateAssignmentType as mxCoordinateAssignment };
/** @type {?} */
const mxHierarchicalLayoutStageType = mxgraph.mxHierarchicalLayoutStage;
export { mxHierarchicalLayoutStageType as mxHierarchicalLayoutStage };
/** @type {?} */
const mxMedianHybridCrossingReductionType = mxgraph.mxMedianHybridCrossingReduction;
export { mxMedianHybridCrossingReductionType as mxMedianHybridCrossingReduction };
/** @type {?} */
const mxMinimumCycleRemoverType = mxgraph.mxMinimumCycleRemover;
export { mxMinimumCycleRemoverType as mxMinimumCycleRemover };
/** @type {?} */
const mxSwimlaneOrderingType = mxgraph.mxSwimlaneOrdering;
export { mxSwimlaneOrderingType as mxSwimlaneOrdering };
