// @edp-bif/runtime-api/esm2015/lib/ref/mxgraph/mxgraphFactory.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ref/mxgraph/mxgraphFactory.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Based on mxgraph 4.2.0
 */
import mxgraphFactory from "mxgraph";
window['mxBasePath'] = window['mxBasePath'] || "/platform/runtime/common/web/@edp-pmf/mxgraph";
window['mxLoadResources'] = window['mxLoadResources'] || undefined;
window['mxForceIncludes'] = window['mxForceIncludes'] || undefined;
window['mxResourceExtension'] = window['mxResourceExtension'] || undefined;
window['mxLoadStylesheets'] = window['mxLoadStylesheets'] || undefined;
window['imageBasePath'] = window['imageBasePath'] || undefined;
window['mxLanguage'] = window['mxLanguage'] || undefined;
window['mxDefaultLanguage'] = window['mxDefaultLanguage'] || undefined;
window['mxLanguages'] = window['mxLanguages'] || undefined;
// declare global {
//     var mxBasePath;
//     var mxLoadResources;
//     var mxForceIncludes;
//     var mxResourceExtension;
//     var mxLoadStylesheets;
//     var mxBasePath;
//     var imageBasePath;
//     var mxLanguage;
//     var mxDefaultLanguage;
//     var mxLanguages;
// }
// declare const require:any;
// const mxgraphFactory = require("/platform/common/web/mxgraph/build.js");
/** @type {?} */
export const mxgraph = mxgraphFactory({});
