// @edp-bif/runtime-api/esm2015/lib/ref/mxgraph/Shape.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ref/mxgraph/Shape.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { mxgraph } from './mxgraphFactory';
/**
 * Shape
 * @type {?}
 */
const mxActorType = mxgraph.mxActor;
export { mxActorType as mxActor };
/** @type {?} */
const mxArrowType = mxgraph.mxArrow;
export { mxArrowType as mxArrow };
/** @type {?} */
const mxArrowConnectorType = mxgraph.mxArrowConnector;
export { mxArrowConnectorType as mxArrowConnector };
/** @type {?} */
const mxCloudType = mxgraph.mxCloud;
export { mxCloudType as mxCloud };
/** @type {?} */
const mxConnectorType = mxgraph.mxConnector;
export { mxConnectorType as mxConnector };
/** @type {?} */
const mxCylinderType = mxgraph.mxCylinder;
export { mxCylinderType as mxCylinder };
/** @type {?} */
const mxDoubleEllipseType = mxgraph.mxDoubleEllipse;
export { mxDoubleEllipseType as mxDoubleEllipse };
/** @type {?} */
const mxEllipseType = mxgraph.mxEllipse;
export { mxEllipseType as mxEllipse };
/** @type {?} */
const mxHexagonType = mxgraph.mxHexagon;
export { mxHexagonType as mxHexagon };
/** @type {?} */
const mxImageShapeType = mxgraph.mxImageShape;
export { mxImageShapeType as mxImageShape };
/** @type {?} */
const mxLabelType = mxgraph.mxLabel;
export { mxLabelType as mxLabel };
/** @type {?} */
const mxLineType = mxgraph.mxLine;
export { mxLineType as mxLine };
/** @type {?} */
const mxMarkerType = mxgraph.mxMarker;
export { mxMarkerType as mxMarker };
/** @type {?} */
const mxPolylineType = mxgraph.mxPolyline;
export { mxPolylineType as mxPolyline };
/** @type {?} */
const mxRectangleShapeType = mxgraph.mxRectangleShape;
export { mxRectangleShapeType as mxRectangleShape };
/** @type {?} */
const mxRhombusType = mxgraph.mxRhombus;
export { mxRhombusType as mxRhombus };
/** @type {?} */
const mxShapeType = mxgraph.mxShape;
export { mxShapeType as mxShape };
/** @type {?} */
const mxStencilType = mxgraph.mxStencil;
export { mxStencilType as mxStencil };
/** @type {?} */
const mxStencilRegistryType = mxgraph.mxStencilRegistry;
export { mxStencilRegistryType as mxStencilRegistry };
/** @type {?} */
const mxSwimlaneType = mxgraph.mxSwimlane;
export { mxSwimlaneType as mxSwimlane };
/** @type {?} */
const mxTextType = mxgraph.mxText;
export { mxTextType as mxText };
/** @type {?} */
const mxTriangleType = mxgraph.mxTriangle;
export { mxTriangleType as mxTriangle };
