// @edp-bif/runtime-api/esm2015/lib/ref/mxgraph/View.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ref/mxgraph/View.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { mxgraph } from './mxgraphFactory';
/**
 * View
 * @type {?}
 */
const mxCellEditorType = mxgraph.mxCellEditor;
export { mxCellEditorType as mxCellEditor };
/** @type {?} */
const mxCellOverlayType = mxgraph.mxCellOverlay;
export { mxCellOverlayType as mxCellOverlay };
/** @type {?} */
const mxCellRendererType = mxgraph.mxCellRenderer;
export { mxCellRendererType as mxCellRenderer };
/** @type {?} */
const mxCellStateType = mxgraph.mxCellState;
export { mxCellStateType as mxCellState };
/** @type {?} */
const mxCellStatePreviewType = mxgraph.mxCellStatePreview;
export { mxCellStatePreviewType as mxCellStatePreview };
/** @type {?} */
const mxConnectionConstraintType = mxgraph.mxConnectionConstraint;
export { mxConnectionConstraintType as mxConnectionConstraint };
/** @type {?} */
const mxEdgeStyleType = mxgraph.mxEdgeStyle;
export { mxEdgeStyleType as mxEdgeStyle };
/** @type {?} */
const mxGraphType = mxgraph.mxGraph;
export { mxGraphType as mxGraph };
/** @type {?} */
const mxGraphSelectionModelType = mxgraph.mxGraphSelectionModel;
export { mxGraphSelectionModelType as mxGraphSelectionModel };
/** @type {?} */
const mxGraphViewType = mxgraph.mxGraphView;
export { mxGraphViewType as mxGraphView };
/** @type {?} */
const mxLayoutManagerType = mxgraph.mxLayoutManager;
export { mxLayoutManagerType as mxLayoutManager };
/** @type {?} */
const mxMultiplicityType = mxgraph.mxMultiplicity;
export { mxMultiplicityType as mxMultiplicity };
/** @type {?} */
const mxOutlineType = mxgraph.mxOutline;
export { mxOutlineType as mxOutline };
/** @type {?} */
const mxPerimeterType = mxgraph.mxPerimeter;
export { mxPerimeterType as mxPerimeter };
/** @type {?} */
const mxPrintPreviewType = mxgraph.mxPrintPreview;
export { mxPrintPreviewType as mxPrintPreview };
/** @type {?} */
const mxStyleRegistryType = mxgraph.mxStyleRegistry;
export { mxStyleRegistryType as mxStyleRegistry };
/** @type {?} */
const mxStylesheetType = mxgraph.mxStylesheet;
export { mxStylesheetType as mxStylesheet };
/** @type {?} */
const mxSwimlaneManagerType = mxgraph.mxSwimlaneManager;
export { mxSwimlaneManagerType as mxSwimlaneManager };
/** @type {?} */
const mxTemporaryCellStatesType = mxgraph.mxTemporaryCellStates;
export { mxTemporaryCellStatesType as mxTemporaryCellStates };
