// @ecp-caf/common-structure/esm2015/lib/convertor/constraint/constraint-convertor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ConstraintConst } from '../../define/constant/constraint-const';
export class ConstraintConvertor {
    /**
     * @param {?} obj
     * @param {?} constraint
     * @param {?} ctxt
     * @return {?}
     */
    convertToObject(obj, constraint, ctxt) {
        if (!constraint) {
            return;
        }
        switch (constraint.getKind()) {
            case ConstraintConst.DECIMAL:
                /** @type {?} */
                let decimalconstraint = (/** @type {?} */ (constraint));
                obj['scale'] = (/** @type {?} */ (decimalconstraint.scale));
                obj['pricision'] = (/** @type {?} */ (decimalconstraint.pricision));
            case ConstraintConst.STRING:
                /** @type {?} */
                let stringconstraint = (/** @type {?} */ (constraint));
                obj['maxLength'] = (/** @type {?} */ (stringconstraint.maxLength));
            default:
                throw new Error(`约束不支持的${constraint.getKind()}类型！`);
        }
    }
    /**
     * @param {?} constraint
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    initFromObject(constraint, obj, ctxt) {
        if (!obj) {
            return null;
        }
        switch (obj['kind']) {
            case ConstraintConst.DECIMAL:
            case ConstraintConst.STRING:
            default:
                throw new Error(`约束不支持的${obj['kind']}类型！`);
        }
    }
}
