// @farris/component-querycondition/esm5/lib/controlcomponent/numbercomponent/single-number.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/controlcomponent/numbercomponent/single-number.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { SolutionService } from '../../service/solution.service';
import { SingleNumberValue } from '../../entity/conditionvalue/single-number-value';
var SingleNumberComponent = /** @class */ (function () {
    function SingleNumberComponent(solutionService) {
        this.solutionService = solutionService;
        this.hideLabel = false;
        /**
         * 某个字段发生值变化事件时，需要同步清空的字段，可能不是当前字段
         */
        this.changeFields = [];
        this.conditionChange = new EventEmitter();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    SingleNumberComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        if (changes['changeFields'] && changes['changeFields'].currentValue && !changes['changeFields'].firstChange) {
            /** @type {?} */
            var isChange = changes['changeFields'].currentValue.some((/**
             * @param {?} field
             * @return {?}
             */
            function (field) { return field == _this.item.fieldCode; }));
            if (isChange) {
                this.number = this.item.value.getOriginalValue();
            }
        }
    };
    /**
     * @return {?}
     */
    SingleNumberComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.number = this.item.value.getOriginalValue();
        this.testId = this.solutionService.getGuid();
        this.canNull = true;
        this.placeholder = (this.item.placeHolder) ? (this.item.placeHolder) : (this.item.control.placeholder);
        this.precision = this.item.control.precision;
        this.textAlign = this.item.control.textAlign;
        this.bigNumber = this.item.control.bigNumber ? this.item.control.bigNumber : false;
        if (!this.bigNumber) {
            this.min = (this.item.control.min === undefined || this.item.control.min === null) ? -2147483648 : (this.item.control.min);
            this.max = (this.item.control.max === undefined || this.item.control.max === null) ? 2147483647 : (this.item.control.max);
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    SingleNumberComponent.prototype.valueChange = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        /** @type {?} */
        var numberValue = new SingleNumberValue();
        isNaN(parseFloat(data)) ? numberValue.numValue = null : numberValue.numValue = data;
        this.item.value = numberValue;
        this.conditionChange.emit();
    };
    // getConditionItemStyle(field: QueryCondition, controlType: any) {
    //     return this.solutionService.getConditionItemStyle(field,controlType);
    // }
    // getConditionItemStyle(field: QueryCondition, controlType: any) {
    //     return this.solutionService.getConditionItemStyle(field,controlType);
    // }
    /**
     * @param {?} field
     * @return {?}
     */
    SingleNumberComponent.prototype.getConditionItemTitle = 
    // getConditionItemStyle(field: QueryCondition, controlType: any) {
    //     return this.solutionService.getConditionItemStyle(field,controlType);
    // }
    /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        return this.solutionService.getConditionItemTitle(field);
    };
    SingleNumberComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-solution-single-number',
                    template: "<div class=\"farris-group-wrap\" id=\"{{item.id}}\" >\r\n    <div class=\"form-group farris-form-group  common-group\" [ngClass]=\"{'q-state-readonly':item.readonly}\">\r\n        <label class=\"col-form-label\" *ngIf=\"item && !hideLabel\" title=\"{{getConditionItemTitle(item)}}\">\r\n            <span class=\"farris-label-info text-danger\" *ngIf=\"item.control.isRequired()\">*</span>\r\n            <span class=\"farris-label-text\">{{item.fieldName}}</span>\r\n        </label>\r\n        <div class=\"farris-input-wrap\">\r\n            <farris-number-spinner *ngIf=\"!bigNumber\"\r\n                [(ngModel)]=\"number\"\r\n                (valueChange)=\"valueChange($event)\"\r\n                id=\"{{testId}}\"\r\n                [min]=\"min\"\r\n                [max]=\"max\"\r\n                [canNull]=\"canNull\"\r\n                [bigNumber]=\"bigNumber\"\r\n                [placeholder]=\"placeholder\"\r\n                [precision]=\"precision\"\r\n                [textAlign]=\"textAlign\"\r\n                [readonly]=\"item.readonly\">\r\n            </farris-number-spinner>\r\n            <farris-number-spinner *ngIf=\"bigNumber\"\r\n                [(ngModel)]=\"number\"\r\n                (valueChange)=\"valueChange($event)\"\r\n                id=\"{{testId}}\"\r\n                [canNull]=\"canNull\"\r\n                [bigNumber]=\"bigNumber\"\r\n                [placeholder]=\"placeholder\"\r\n                [precision]=\"precision\"\r\n                [textAlign]=\"textAlign\"\r\n                [readonly]=\"item.readonly\">\r\n            </farris-number-spinner>\r\n        </div>\r\n    </div>\r\n</div>"
                }] }
    ];
    /** @nocollapse */
    SingleNumberComponent.ctorParameters = function () { return [
        { type: SolutionService }
    ]; };
    SingleNumberComponent.propDecorators = {
        hideLabel: [{ type: Input }],
        item: [{ type: Input }],
        changeFields: [{ type: Input }],
        conditionChange: [{ type: Output, args: ['conditionChange',] }]
    };
    return SingleNumberComponent;
}());
export { SingleNumberComponent };
if (false) {
    /** @type {?} */
    SingleNumberComponent.prototype.hideLabel;
    /** @type {?} */
    SingleNumberComponent.prototype.number;
    /** @type {?} */
    SingleNumberComponent.prototype.placeholder;
    /** @type {?} */
    SingleNumberComponent.prototype.precision;
    /** @type {?} */
    SingleNumberComponent.prototype.textAlign;
    /** @type {?} */
    SingleNumberComponent.prototype.item;
    /** @type {?} */
    SingleNumberComponent.prototype.testId;
    /** @type {?} */
    SingleNumberComponent.prototype.canNull;
    /** @type {?} */
    SingleNumberComponent.prototype.bigNumber;
    /** @type {?} */
    SingleNumberComponent.prototype.min;
    /** @type {?} */
    SingleNumberComponent.prototype.max;
    /**
     * 某个字段发生值变化事件时，需要同步清空的字段，可能不是当前字段
     * @type {?}
     */
    SingleNumberComponent.prototype.changeFields;
    /** @type {?} */
    SingleNumberComponent.prototype.conditionChange;
    /**
     * @type {?}
     * @private
     */
    SingleNumberComponent.prototype.solutionService;
}
