// @farris/ui-organization-selector/esm2015/lib/tab-favorites/favorites.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/tab-favorites/favorites.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectorRef, Injector } from '@angular/core';
import * as _ from 'lodash-es';
import { noDataImg } from '../../assets/resource';
import { ApiService } from '../../utils/api.service';
import { CommonUtil } from '../../utils/common.tool';
import { FavoritesTool } from '../../utils/favorites.tool';
import { ChangeType } from '../../utils/index.type';
import { StateMgrTool } from '../../utils/state.mgr.tool';
import { LoadingService } from '@farris/ui-loading';
export class FavoritesComponent {
    /**
     * @param {?} apiService
     * @param {?} favoritesTool
     * @param {?} stateMgr
     * @param {?} cdr
     * @param {?} injector
     */
    constructor(apiService, favoritesTool, stateMgr, cdr, injector) {
        this.apiService = apiService;
        this.favoritesTool = favoritesTool;
        this.stateMgr = stateMgr;
        this.cdr = cdr;
        this.injector = injector;
        this.disableOrgList = [];
        this.allChecked = false;
        this.selections = [];
        this.allIndexId = 'allIndexId';
        this.noDataImg = noDataImg;
        this.items = [];
        this.favoritesList = [];
        this.loadingService = null;
        this.allIndexId = CommonUtil.newGuid();
        this.loadingService = this.injector.get(LoadingService, null);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.selectionsSub.unsubscribe();
        this.favoritesSub.unsubscribe();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.loadingService) {
            this.loadingService.show();
        }
        this.favoritesSub = this.favoritesTool.favoritesSubject.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            /** @type {?} */
            let resData = _.cloneDeep(data);
            this.favoritesList = resData;
            this.loadFavoritesList();
        }));
        this.selectionsSub = this.stateMgr.selectionsSubject.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            if (data.changeType === ChangeType.ValueChanged) {
                this.selections = _.cloneDeep(data.value);
            }
            else if (data.changeType === ChangeType.Remove) {
                this.unSelectLine(data.value);
            }
            else if (data.changeType === ChangeType.Append) {
                this.selectLine(data.value);
            }
            this.toggleChecked();
            this.cdr.detectChanges();
        }));
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
    }
    /**
     * @return {?}
     */
    assembleData() {
        this.favoritesSub = this.favoritesTool.favoritesSubject.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            this.favoritesList = data;
            this.loadFavoritesList();
        }));
    }
    /**
     * @param {?} e
     * @param {?} item
     * @return {?}
     */
    deleteFromFavorites(e, item) {
        this.favoritesTool.toggleFavorites(e, item);
    }
    /**
     * @param {?} e
     * @return {?}
     */
    toggleAll(e) {
        if (!this.allChecked) {
            this.checkAll();
        }
        else {
            this.unCheckAll();
        }
        e.preventDefault();
    }
    /**
     * @param {?} data
     * @param {?} e
     * @return {?}
     */
    toggleOne(data, e) {
        if (!data.enable) {
            e.stopPropagation();
            e.cancelBubble = true;
            return;
        }
        if (data.checked) {
            if (this.multiSelect) {
                this.emitSelections(data, ChangeType.Remove);
            }
            else {
                this.emitSelections([]);
            }
        }
        else {
            if (this.multiSelect) {
                this.emitSelections(data, ChangeType.Append);
            }
            else {
                this.emitSelections(new Array(data));
            }
        }
        e.preventDefault();
    }
    /**
     * 用于数据展示
     * @return {?}
     */
    loadFavoritesList() {
        this.items = [];
        this.favoritesList.forEach((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            /** @type {?} */
            const indexId = CommonUtil.newGuid();
            /** @type {?} */
            const item = {
                orgId: res.orgId,
                code: res.code,
                name: res.name ? res.name : '',
                orgAllPath: !res.orgAllPath || res.orgAllPath.split('_').includes('null') ? '' : res.orgAllPath,
                path: res.path,
                layer: Number(res.layer),
                enable: this.filterType == 'disable' ? !this.disableOrgList.includes(res.orgId) : true,
                indexId
            };
            if (CommonUtil.findIndex(res[this.primaryField], this.primaryField, this.selections) !== -1) {
                item.checked = true;
            }
            this.items.push(item);
        }));
        this.items = this.items.slice();
        this.cdr.detectChanges();
        if (this.loadingService) {
            this.loadingService.close();
        }
    }
    /**
     * 全选后处理
     * @return {?}
     */
    checkAll() {
        this.items.forEach((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            if (CommonUtil.findIndex(item[this.primaryField], this.primaryField, this.selections) === -1 && item.enable) {
                this.emitSelections(item, ChangeType.Append);
            }
        }));
    }
    /**
     * 全部取消
     * @return {?} void
     */
    unCheckAll() {
        /** @type {?} */
        const checkeds = this.items.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => item.checked));
        checkeds.forEach((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            this.emitSelections(item, ChangeType.Remove);
        }));
    }
    /**
     * 当前行选中
     * @private
     * @param {?} data 当前行的数据（）
     * @return {?}
     */
    selectLine(data) {
        if (this.multiSelect) {
            this._insertOneIntoSelections(data);
        }
        else {
            this.selections = [];
            this.items.forEach((/**
             * @param {?} item
             * @return {?}
             */
            item => {
                if (item.checked) {
                    item.checked = false;
                }
            }));
            this._insertOneIntoSelections(data);
        }
        if (this.isAllChecked()) {
            this.allChecked = true;
        }
    }
    /**
     * 单选取消选中
     * @private
     * @param {?} data 取消选中行
     * @return {?}
     */
    unSelectLine(data) {
        /** @type {?} */
        const index = CommonUtil.findIndex(data[this.primaryField], this.primaryField, this.selections);
        if (index > -1) {
            this.selections.splice(index, 1);
        }
        this.setChecked(data, false);
        if (this.allChecked) {
            this.allChecked = false;
        }
    }
    /**
     * @private
     * @param {?} data
     * @param {?} checked
     * @return {?}
     */
    setChecked(data, checked) {
        /** @type {?} */
        const itemIndex = CommonUtil.findIndex(data[this.primaryField], this.primaryField, this.items);
        if (itemIndex > -1) {
            this.items[itemIndex].checked = checked;
        }
    }
    /**
     * @private
     * @param {?} selections
     * @param {?=} changeType
     * @param {?=} from
     * @return {?}
     */
    emitSelections(selections, changeType = ChangeType.ValueChanged, from = 'favorite') {
        this.stateMgr.emitSelections(selections, changeType, from);
    }
    /**
     * @param {?} e
     * @param {?} item
     * @return {?}
     */
    toggleFavorites(e, item) {
        this.favoritesTool.toggleFavorites(e, item);
    }
    /**
     * 全部取消勾选
     * @return {?}
     */
    clearCheckeds() {
        this.items.forEach((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            item.checked = false;
        }));
        this.allChecked = false;
        this.selections = [];
    }
    /**
     * 选中并插入当前行数据到右侧
     * @param {?} data 当前行的数据
     * @return {?}
     */
    _insertOneIntoSelections(data) {
        // 如果当前数据已经移入右侧数据，则返回。
        if (CommonUtil.findIndex(data[this.primaryField], this.primaryField, this.selections) > -1) {
            return;
        }
        this.setChecked(data, true);
        this.selections.push(data);
        this.selections = this.selections.slice();
    }
    /**
     * 翻页检查：人员展示区的勾选状态与已选人员区是否对应
     * @return {?}
     */
    toggleChecked() {
        this.items.forEach((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            if (CommonUtil.findIndex(item[this.primaryField], this.primaryField, this.selections) > -1 && item.enable) {
                item.checked = true;
            }
            else {
                item.checked = false;
            }
        }));
        if (this.isAllChecked()) {
            this.allChecked = true;
        }
        else {
            this.allChecked = false;
        }
    }
    /**
     * @private
     * @return {?}
     */
    isAllChecked() {
        /** @type {?} */
        let count = 0;
        /** @type {?} */
        let allChecked = false;
        this.items.forEach((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            this.selections.forEach((/**
             * @param {?} selection
             * @return {?}
             */
            selection => {
                if (item[this.primaryField] === selection[this.primaryField]) {
                    count += 1;
                }
            }));
        }));
        if (this.items.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => item.enable)).length === count && count > 0) {
            allChecked = true;
        }
        return allChecked;
    }
    /**
     * @param {?} index
     * @param {?} item
     * @return {?}
     */
    trackByFn(index, item) {
        return index;
    }
}
FavoritesComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-favorites',
                template: "<div class=\"os-card\">\r\n    <div class=\"os-card-title\" *ngIf=\"multiSelect && items.length > 0\">\r\n        <div class=\"custom-control custom-checkbox\">\r\n            <input class=\"custom-control-input\" [id]=\"allIndexId\" type=\"checkbox\" [checked]=\"allChecked\">\r\n            <label class=\"custom-control-label\" [for]=\"allIndexId\"\r\n                (click)=\"toggleAll($event)\">{{'organizationSelector.tabCommon.firstCheckAllText' | locale}}{{items.length}}{{'organizationSelector.tabCommon.lastCheckAllText' | locale}}</label>\r\n        </div>\r\n    </div>\r\n    <div class=\"os-card-blank\" *ngIf=\"!multiSelect\"></div>\r\n    <perfect-scrollbar>\r\n        <div class=\"os-card-content\">\r\n            <div *ngFor=\"let data of items; let i = index;trackBy: trackByFn\" class=\"s-selectedPerson\" [ngClass]=\"{'disabledPerson': !data.enable}\">\r\n                <div class=\"custom-control custom-checkbox mb-2\">\r\n                    <input class=\"custom-control-input\" [id]=\"data.indexId\" type=\"checkbox\" [disabled]=\"!data.enable\" [checked]=\"data.checked\">\r\n                    <label class=\"custom-control-label\" [for]=\"data.indexId\" (click)=\"toggleOne(data, $event)\">\r\n                        <div class=\"src-item-container\">\r\n                            <div class=\"src-item-content\">\r\n                                <p class=\"os-card-content-title\" title=\"{{data.name}}\">\r\n                                    {{data.name}}</p>\r\n                                <p class=\"os-card-content-orgpath\" title=\"{{data.orgAllPath}}\">\r\n                                    {{data.orgAllPath}}</p>\r\n                            </div>\r\n                        </div>\r\n                    </label>\r\n                </div>\r\n                <span class=\"f-icon f-icon-minus-circle\"\r\n                    (click)=\"deleteFromFavorites($event, data)\"></span>\r\n            </div>\r\n            <div *ngIf=\"items.length===0\" [ngStyle]=\"{'background-image': noDataImg}\" class=\"fp-os-card-nodata\">\r\n                {{'organizationSelector.tabCommon.noOrgData' | locale}}</div>\r\n        </div>\r\n    </perfect-scrollbar>\r\n    <!-- \u5206\u9875\r\n    <div class=\"pagination-container\" *ngIf=\"items.length!==0\">\r\n        <ul class=\"ngx-pagination pagination pager-viewmode-simple\">\r\n            <li class=\"page-item d-flex flex-fill\"></li>\r\n            <li class=\"page-item\" [class.disabled]=\"pageIndex === 1\" (click)=\"turnOverPage('pre', -1)\">\r\n                <span class=\"page-link border-0\">\r\n                    <span class=\"f-icon f-page-pre\"></span>\r\n                </span>\r\n            </li>\r\n            <li class=\"page-item\">\r\n                <input id=\"farris-datagrid-gotopagenumber\" min=\"1\" type=\"number\" max=\"{{pageTotal}}\"\r\n                    [(ngModel)]=\"pageIndex\" style=\"width: 38px;\" (keyup.enter)=\"switchPage({pageIndex: pageIndex})\"\r\n                    (input)=\"beforePageIndexChange($event, pageTotal)\">\r\n            </li>\r\n            <li class=\"page-item page-separator\" style=\"margin-left: 10px\"> / </li>\r\n            <li class=\"page-item page-total\" style=\"margin-left: 5px\">\r\n                <span style=\"font-size: 16px; font-weight: 600;\"> {{pageTotal}}</span>\r\n            </li>\r\n            <li class=\"page-item\" [class.disabled]=\"pageIndex === pageTotal\" (click)=\"turnOverPage('next', 1)\">\r\n                <a class=\"page-link border-0\">\r\n                    <span class=\"f-icon f-page-next\"></span>\r\n                </a>\r\n            </li>\r\n        </ul>\r\n    </div>\r\n    -->\r\n</div>",
                styles: [":host{background:#fff;display:-webkit-box;display:flex;width:100%;height:100%}.os-card{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;-webkit-box-flex:1;flex:1;overflow:hidden}.os-card .s-selectedPerson:hover{background:#eef5ff;border-radius:5px}.os-card .os-card-content{padding-top:3px;clear:both}.os-card .os-card-content .src-item-container{display:-webkit-box;display:flex;height:32px;margin:-21px 0 0 15px}.os-card .os-card-content .src-item-content{margin-left:11px;margin-top:-2px;padding-right:10px;overflow:hidden}.os-card .s-selectedPerson{min-width:240px;float:left;width:calc(50% - 6px);position:relative;height:40px;padding:5px 12px;margin:0 3px 10px}.os-card .disabledPerson{color:#a8adb8;background-color:#fafafc}.os-card .disabledPerson:hover{background:#fafafc!important}.os-card .s-selectedPerson .f-icon-minus-circle,.os-card .s-selectedPerson .f-icon-star{position:absolute;top:13px;right:15px;font-size:14px;cursor:pointer}.os-card .s-selectedPerson .f-icon-minus-circle{color:#f46160}.os-card .s-selectedPerson .default-light-color{color:#f7b500}.os-card .s-selectedPerson .default-dark-color{color:#d8dce6}.os-card .k-checkbox-label::after,.os-card .k-checkbox-label::before{top:10px;font-size:16px;width:14px;height:14px}.os-card .fp-os-card-nodata{width:100px;height:100px;text-align:center;background-repeat:no-repeat;background-size:contain;color:#b4bccc;padding-top:110px;position:absolute;left:50%;top:50%;margin:-70px 0 0 -50px}.os-card .os-card-content-title{margin:0;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;font-size:13px}.os-card .os-card-content-orgpath{margin:0;font-size:12px;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;color:#c7ccd7}.os-card .os-card-title{display:-webkit-box;display:flex;-webkit-box-pack:justify;justify-content:space-between;font-size:14px}.os-card .os-card-title .custom-control{margin-left:15px;line-height:35px}.os-card .os-card-title .title-clearall{margin:10px 15px 0 0;color:#2a87ff;cursor:pointer;line-height:35px}.os-card .s-selectedPerson .custom-control-label::before{top:1px}.os-card .os-card-title .f-icon{color:#878d99;cursor:pointer}.os-card .os-card-blank{width:100%;height:20px}"]
            }] }
];
/** @nocollapse */
FavoritesComponent.ctorParameters = () => [
    { type: ApiService },
    { type: FavoritesTool },
    { type: StateMgrTool },
    { type: ChangeDetectorRef },
    { type: Injector }
];
FavoritesComponent.propDecorators = {
    primaryField: [{ type: Input }],
    multiSelect: [{ type: Input }],
    filterType: [{ type: Input }],
    disableOrgList: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FavoritesComponent.prototype.primaryField;
    /** @type {?} */
    FavoritesComponent.prototype.multiSelect;
    /** @type {?} */
    FavoritesComponent.prototype.filterType;
    /** @type {?} */
    FavoritesComponent.prototype.disableOrgList;
    /** @type {?} */
    FavoritesComponent.prototype.allChecked;
    /** @type {?} */
    FavoritesComponent.prototype.selections;
    /** @type {?} */
    FavoritesComponent.prototype.allIndexId;
    /** @type {?} */
    FavoritesComponent.prototype.noDataImg;
    /** @type {?} */
    FavoritesComponent.prototype.items;
    /** @type {?} */
    FavoritesComponent.prototype.favoritesList;
    /**
     * @type {?}
     * @private
     */
    FavoritesComponent.prototype.loadingService;
    /**
     * @type {?}
     * @private
     */
    FavoritesComponent.prototype.selectionsSub;
    /**
     * @type {?}
     * @private
     */
    FavoritesComponent.prototype.favoritesSub;
    /**
     * @type {?}
     * @private
     */
    FavoritesComponent.prototype.apiService;
    /**
     * @type {?}
     * @private
     */
    FavoritesComponent.prototype.favoritesTool;
    /**
     * @type {?}
     * @private
     */
    FavoritesComponent.prototype.stateMgr;
    /**
     * @type {?}
     * @private
     */
    FavoritesComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    FavoritesComponent.prototype.injector;
}
