// @farris/ide-devkit/esm5/lib/components/angular-editor/editor.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ANGULAR_MONACO_EDITOR_CONFIG } from './model/config';
import { AngularMonacoEditorComponent } from './angular-monaco-editor/angular-monaco-editor.component';
var AngularMonacoEditorModule = /** @class */ (function () {
    function AngularMonacoEditorModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    AngularMonacoEditorModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: AngularMonacoEditorModule,
            providers: [
                { provide: ANGULAR_MONACO_EDITOR_CONFIG, useValue: config }
            ]
        };
    };
    AngularMonacoEditorModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        AngularMonacoEditorComponent
                    ],
                    exports: [
                        AngularMonacoEditorComponent
                    ],
                    providers: []
                },] }
    ];
    return AngularMonacoEditorModule;
}());
export { AngularMonacoEditorModule };
