// @farris/ide-devkit/esm2015/lib/components/dropdown/dropdown-button.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostBinding, HostListener, Input } from '@angular/core';
import { BsDropdownState } from './dropdown.state';
export class DropdownButtonDirective {
    /**
     * @param {?} _state
     * @param {?} _element
     */
    constructor(_state, _element) {
        this._state = _state;
        this._element = _element;
        this._subscriptions = [];
        this.isDisabled = null;
        // 同步菜单的打开状态
        this._subscriptions.push(this._state.isOpenChange
            .subscribe((value) => this.isOpen = value));
        // 同步菜单的禁用状态
        this._subscriptions.push(this._state.isDisabledChange
            .subscribe((value) => this.isDisabled = value || null));
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onClick(event) {
        if (this.isDisabled) {
            return;
        }
        this._state.onClick.emit(event);
    }
    /**
     * @return {?}
     */
    onHover() {
        if (this.isDisabled) {
            return;
        }
        this._state.onHover.emit();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        for (const sub of this._subscriptions) {
            sub.unsubscribe();
        }
    }
}
DropdownButtonDirective.decorators = [
    { type: Directive, args: [{
                selector: '[dropdownButton]'
            },] }
];
/** @nocollapse */
DropdownButtonDirective.ctorParameters = () => [
    { type: BsDropdownState },
    { type: ElementRef }
];
DropdownButtonDirective.propDecorators = {
    isDisabled: [{ type: HostBinding, args: ['class.disabled',] }, { type: Input }],
    isOpen: [{ type: HostBinding, args: ['attr.aria-expanded',] }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }],
    onHover: [{ type: HostListener, args: ['mouseover',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    DropdownButtonDirective.prototype._subscriptions;
    /** @type {?} */
    DropdownButtonDirective.prototype.isDisabled;
    /** @type {?} */
    DropdownButtonDirective.prototype.isOpen;
    /**
     * @type {?}
     * @private
     */
    DropdownButtonDirective.prototype._state;
    /**
     * @type {?}
     * @private
     */
    DropdownButtonDirective.prototype._element;
}
