// @farris/ui-calendar/esm5/lib/calendar.locale.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { LocaleService } from '@farris/ui-locale';
import * as i0 from "@angular/core";
import * as i1 from "@farris/ui-locale";
var CalendarLocaleService = /** @class */ (function () {
    function CalendarLocaleService(localeService) {
        this.localeService = localeService;
        this.localeConfig = {
            'ZH_CN': {
                locale: 'zh-cn',
                buttonText: {
                    today: '今天',
                    month: '月',
                    week: '周',
                    day: '天',
                    listWeek: '列表'
                },
                allDayText: '全天',
                noEventsMessage: '没数据啊',
                eventLimitText: '更多',
                weekNumberTitle: '周',
                firstDay: 1
            },
            'EN_US': {
                locale: 'en'
            }
        };
    }
    /**
     * @return {?}
     */
    CalendarLocaleService.prototype.getLocaleConfig = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var locale = this.localeService.getValue('locale');
        if (locale && this.localeConfig[locale]) {
            return this.localeConfig[locale];
        }
        else {
            return this.localeConfig.ZH_CN;
        }
    };
    CalendarLocaleService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    CalendarLocaleService.ctorParameters = function () { return [
        { type: LocaleService }
    ]; };
    /** @nocollapse */ CalendarLocaleService.ngInjectableDef = i0.defineInjectable({ factory: function CalendarLocaleService_Factory() { return new CalendarLocaleService(i0.inject(i1.LocaleService)); }, token: CalendarLocaleService, providedIn: "root" });
    return CalendarLocaleService;
}());
export { CalendarLocaleService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CalendarLocaleService.prototype.localeConfig;
    /**
     * @type {?}
     * @private
     */
    CalendarLocaleService.prototype.localeService;
}
