// @farris/ui-calendar/esm5/lib/calendar.config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import dayGridPlugin from '@fullcalendar/daygrid';
import timeGrigPlugin from '@fullcalendar/timegrid';
import interactionPlugin from '@fullcalendar/interaction';
import { CalendarLocaleService } from './calendar.locale.service';
import * as i0 from "@angular/core";
import * as i1 from "./calendar.locale.service";
var DefaultConfigService = /** @class */ (function () {
    function DefaultConfigService(localeService) {
        this.localeService = localeService;
        this.defaultConfig = {
            header: {
                left: 'prevYear,prev,next,nextYear today',
                center: 'title',
                right: 'dayGridMonth,timeGridWeek,timeGridDay,listWeek'
            },
            navLinks: true,
            // can click day/week names to navigate views
            editable: true,
            selectable: true,
            defaultView: 'dayGridMonth',
            eventLimit: true,
            // allow "more" link when too many events
            plugins: [dayGridPlugin, timeGrigPlugin, interactionPlugin]
        };
    }
    /**
     * @return {?}
     */
    DefaultConfigService.prototype.getDefaultConfig = /**
     * @return {?}
     */
    function () {
        return Object.assign(this.defaultConfig, this.localeService.getLocaleConfig());
        // return this.defaultConfig;
    };
    DefaultConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    DefaultConfigService.ctorParameters = function () { return [
        { type: CalendarLocaleService }
    ]; };
    /** @nocollapse */ DefaultConfigService.ngInjectableDef = i0.defineInjectable({ factory: function DefaultConfigService_Factory() { return new DefaultConfigService(i0.inject(i1.CalendarLocaleService)); }, token: DefaultConfigService, providedIn: "root" });
    return DefaultConfigService;
}());
export { DefaultConfigService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DefaultConfigService.prototype.defaultConfig;
    /**
     * @type {?}
     * @private
     */
    DefaultConfigService.prototype.localeService;
}
