// @farris/ui-perfect-scrollbar/esm2015/lib/perfect-scrollbar.interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/** @type {?} */
export const PERFECT_SCROLLBAR_CONFIG = new InjectionToken('PERFECT_SCROLLBAR_CONFIG');
export class Geometry {
    /**
     * @param {?} x
     * @param {?} y
     * @param {?} w
     * @param {?} h
     */
    constructor(x, y, w, h) {
        this.x = x;
        this.y = y;
        this.w = w;
        this.h = h;
    }
}
if (false) {
    /** @type {?} */
    Geometry.prototype.x;
    /** @type {?} */
    Geometry.prototype.y;
    /** @type {?} */
    Geometry.prototype.w;
    /** @type {?} */
    Geometry.prototype.h;
}
export class Position {
    /**
     * @param {?} x
     * @param {?} y
     */
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
}
if (false) {
    /** @type {?} */
    Position.prototype.x;
    /** @type {?} */
    Position.prototype.y;
}
/** @type {?} */
export const PerfectScrollbarEvents = [
    'psScrollY',
    'psScrollX',
    'psScrollUp',
    'psScrollDown',
    'psScrollLeft',
    'psScrollRight',
    'psYReachEnd',
    'psYReachStart',
    'psXReachEnd',
    'psXReachStart'
];
/**
 * @record
 */
export function PerfectScrollbarConfigInterface() { }
if (false) {
    /** @type {?|undefined} */
    PerfectScrollbarConfigInterface.prototype.handlers;
    /** @type {?|undefined} */
    PerfectScrollbarConfigInterface.prototype.wheelSpeed;
    /** @type {?|undefined} */
    PerfectScrollbarConfigInterface.prototype.swipeEasing;
    /** @type {?|undefined} */
    PerfectScrollbarConfigInterface.prototype.suppressScrollX;
    /** @type {?|undefined} */
    PerfectScrollbarConfigInterface.prototype.suppressScrollY;
    /** @type {?|undefined} */
    PerfectScrollbarConfigInterface.prototype.wheelPropagation;
    /** @type {?|undefined} */
    PerfectScrollbarConfigInterface.prototype.useBothWheelAxes;
    /** @type {?|undefined} */
    PerfectScrollbarConfigInterface.prototype.scrollingThreshold;
    /** @type {?|undefined} */
    PerfectScrollbarConfigInterface.prototype.minScrollbarLength;
    /** @type {?|undefined} */
    PerfectScrollbarConfigInterface.prototype.maxScrollbarLength;
    /** @type {?|undefined} */
    PerfectScrollbarConfigInterface.prototype.scrollXMarginOffset;
    /** @type {?|undefined} */
    PerfectScrollbarConfigInterface.prototype.scrollYMarginOffset;
}
export class PerfectScrollbarConfig {
    /**
     * @param {?=} config
     */
    constructor(config = {}) {
        this.assign(config);
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    assign(config = {}) {
        for (const key in config) {
            this[(/** @type {?} */ (key))] = config[(/** @type {?} */ (key))];
        }
    }
}
if (false) {
    /** @type {?} */
    PerfectScrollbarConfig.prototype.handlers;
    /** @type {?} */
    PerfectScrollbarConfig.prototype.wheelSpeed;
    /** @type {?} */
    PerfectScrollbarConfig.prototype.swipeEasing;
    /** @type {?} */
    PerfectScrollbarConfig.prototype.suppressScrollX;
    /** @type {?} */
    PerfectScrollbarConfig.prototype.suppressScrollY;
    /** @type {?} */
    PerfectScrollbarConfig.prototype.wheelPropagation;
    /** @type {?} */
    PerfectScrollbarConfig.prototype.useBothWheelAxes;
    /** @type {?} */
    PerfectScrollbarConfig.prototype.scrollingThreshold;
    /** @type {?} */
    PerfectScrollbarConfig.prototype.minScrollbarLength;
    /** @type {?} */
    PerfectScrollbarConfig.prototype.maxScrollbarLength;
    /** @type {?} */
    PerfectScrollbarConfig.prototype.scrollXMarginOffset;
    /** @type {?} */
    PerfectScrollbarConfig.prototype.scrollYMarginOffset;
}
