// @gsp-bef/gsp-be-metadata/esm2015/lib/biz-entity/src/Operation/ParameterEnum/BizParameterType.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const BizParameterType = {
    /**
     * 字符型
     */
    String: 0,
    /**
     * 布尔型
     */
    Boolean: 1,
    /**
     * 整数
     */
    Int32: 2,
    /**
     * 浮点数
     */
    Decimal: 3,
    /**
     * 双浮点
     */
    Double: 4,
    /**
     * 时间
     */
    DateTime: 5,
    /**
     * 对象类型
     */
    Object: 6,
    /**
     * 自定义
     */
    Custom: 7,
};
export { BizParameterType };
BizParameterType[BizParameterType.String] = 'String';
BizParameterType[BizParameterType.Boolean] = 'Boolean';
BizParameterType[BizParameterType.Int32] = 'Int32';
BizParameterType[BizParameterType.Decimal] = 'Decimal';
BizParameterType[BizParameterType.Double] = 'Double';
BizParameterType[BizParameterType.DateTime] = 'DateTime';
BizParameterType[BizParameterType.Object] = 'Object';
BizParameterType[BizParameterType.Custom] = 'Custom';
// export const BizParameterTypes: Array<{ index: number, text: string, value: BizParameterType, assembly: string, className: string }> = [
//     { index: 0, text: '字符串', value: BizParameterType.String, assembly: '', className: '' },
//     { index: 1, text: '布尔型', value: BizParameterType.Boolean, assembly: '', className: '' },
//     { index: 2, text: '整数', value: BizParameterType.Integer, assembly: '', className: '' },
//     { index: 3, text: '浮点数', value: BizParameterType.Float, assembly: '', className: '' },
//     { index: 4, text: '双浮点数', value: BizParameterType.Double, assembly: '', className: '' },
//     { index: 5, text: '时间类型', value: BizParameterType.DateTime, assembly: '', className: '' },
//     { index: 6, text: '对象类型', value: BizParameterType.Object, assembly: '', className: '' },
//     { index: 7, text: '自定义类型', value: BizParameterType.Custom, assembly: '', className: '' },
// ];
/** @type {?} */
export const BizParameterTypes = [
    { index: 0, text: '字符串', value: BizParameterType.String, assembly: 'mscorlib.dll', className: 'System.String', javaClassName: 'java.lang.String' },
    { index: 1, text: '布尔型', value: BizParameterType.Boolean, assembly: 'mscorlib.dll', className: 'System.Boolean', javaClassName: 'java.lang.Boolean' },
    { index: 2, text: '整数', value: BizParameterType.Int32, assembly: 'mscorlib.dll', className: 'System.Int32', javaClassName: 'java.lang.Integer' },
    { index: 3, text: '浮点数', value: BizParameterType.Decimal, assembly: 'mscorlib.dll', className: 'System.Single', javaClassName: 'java.math.BigDecimal' },
    { index: 4, text: '双浮点数', value: BizParameterType.Double, assembly: 'mscorlib.dll', className: 'System.Double', javaClassName: 'java.lang.Double' },
    { index: 5, text: '时间类型', value: BizParameterType.DateTime, assembly: 'mscorlib.dll', className: 'System.DateTime', javaClassName: 'java.util.Date' },
    { index: 6, text: '对象类型', value: BizParameterType.Object, assembly: 'mscorlib.dll', className: 'System.Object', javaClassName: 'java.lang.Object' },
    { index: 7, text: '自定义类型', value: BizParameterType.Custom, assembly: '', className: '', javaClassName: '' },
];
