// @gsp-bef/gsp-vo-metadata/bundles/gsp-bef-gsp-vo-metadata.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@gsp-bef/gsp-cm-metadata')) :
    typeof define === 'function' && define.amd ? define('@gsp-bef/gsp-vo-metadata', ['exports', '@gsp-bef/gsp-cm-metadata'], factory) :
    (factory((global['gsp-bef'] = global['gsp-bef'] || {}, global['gsp-bef']['gsp-vo-metadata'] = {}),global.gspCmMetadata));
}(this, (function (exports,gspCmMetadata) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 视图对象类
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 视图对象类
     */
    var  /**
     * 视图对象类
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * 视图对象类
     */
    GSPViewModel = /** @class */ (function (_super) {
        __extends(GSPViewModel, _super);
        function GSPViewModel() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Object.defineProperty(GSPViewModel.prototype, "ExtendType", {
            /**
             * 获取扩展类型,总是返回为'视图对象'类型
             */
            get: /**
             * 获取扩展类型,总是返回为'视图对象'类型
             * @return {?}
             */ function () {
                return 'GSPViewModel';
            },
            enumerable: true,
            configurable: true
        });
        return GSPViewModel;
    }(gspCmMetadata.GSPCommonModel));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * VO节点元素
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * VO节点元素
     */
    var  /**
     * VO节点元素
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * VO节点元素
     */
    GSPViewModelElement = /** @class */ (function (_super) {
        __extends(GSPViewModelElement, _super);
        function GSPViewModelElement() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return GSPViewModelElement;
    }(gspCmMetadata.GSPCommonElement));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * BE节点实体类
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * BE节点实体类
     */
    var  /**
     * BE节点实体类
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * BE节点实体类
     */
    GSPViewObject = /** @class */ (function (_super) {
        __extends(GSPViewObject, _super);
        function GSPViewObject() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return GSPViewObject;
    }(gspCmMetadata.GSPCommonObject));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 数据逻辑扩展
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 数据逻辑扩展
     */
    var  /**
     * 数据逻辑扩展
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * 数据逻辑扩展
     */
    VoDataExtendInfo = /** @class */ (function () {
        function VoDataExtendInfo() {
        }
        return VoDataExtendInfo;
    }());
    /** @type {?} */
    var VoDataExtendInfos = [{
            Name: '扩展操作',
            Value: 'DataExtendInfo',
            items: [
                {
                    Index: 0,
                    Name: '数据Mapping',
                    Value: 'DataMappingActions',
                    HasInternalAction: true
                },
                {
                    Index: 1,
                    Name: '查询数据前',
                    Value: 'BeforeQueryActions',
                    HasInternalAction: false
                },
                {
                    Index: 2,
                    Name: '查询数据',
                    Value: 'QueryActions',
                    HasInternalAction: true
                },
                {
                    Index: 3,
                    Name: '查询数据后',
                    Value: 'AfterQueryActions',
                    HasInternalAction: false
                },
                {
                    Index: 4,
                    Name: '检索数据前',
                    Value: 'BeforeRetrieveActions',
                    HasInternalAction: false
                },
                {
                    Index: 5,
                    Name: '检索数据',
                    Value: 'RetrieveActions',
                    HasInternalAction: true
                },
                {
                    Index: 6,
                    Name: '检索数据后',
                    Value: 'AfterRetrieveActions',
                    HasInternalAction: false
                },
                {
                    Index: 7,
                    Name: '修改数据前',
                    Value: 'BeforeModifyActions',
                    HasInternalAction: false
                },
                {
                    Index: 8,
                    Name: '修改数据',
                    Value: 'ModifyActions',
                    HasInternalAction: true
                },
                {
                    Index: 9,
                    Name: '修改数据后',
                    Value: 'AfterModifyActions',
                    HasInternalAction: false
                },
                {
                    Index: 10,
                    Name: '变更集Mapping',
                    Value: 'ChangesetMappingActions',
                    HasInternalAction: true
                },
                {
                    Index: 11,
                    Name: '新增数据前',
                    Value: 'BeforeCreateActions',
                    HasInternalAction: false
                },
                {
                    Index: 12,
                    Name: '新增数据',
                    Value: 'CreateActions',
                    HasInternalAction: true
                },
                {
                    Index: 13,
                    Name: '新增数据后',
                    Value: 'AfterCreateActions',
                    HasInternalAction: false
                },
                {
                    Index: 14,
                    Name: '删除数据前',
                    Value: 'BeforeDeleteActions',
                    HasInternalAction: false
                },
                {
                    Index: 15,
                    Name: '删除数据',
                    Value: 'DeleteActions',
                    HasInternalAction: true
                },
                {
                    Index: 16,
                    Name: '删除数据后',
                    Value: 'AfterDeleteActions',
                    HasInternalAction: false
                },
                {
                    Index: 17,
                    Name: '批量删除数据前',
                    Value: 'BeforeMultiDeleteActions',
                    HasInternalAction: false
                },
                {
                    Index: 18,
                    Name: '批量删除数据',
                    Value: 'MultiDeleteActions',
                    HasInternalAction: true
                },
                {
                    Index: 19,
                    Name: '批量删除数据后',
                    Value: 'AfterMultiDeleteActions',
                    HasInternalAction: false
                },
                {
                    Index: 20,
                    Name: '保存数据前',
                    Value: 'BeforeSaveActions',
                    HasInternalAction: false
                },
                {
                    Index: 21,
                    Name: '数据反向Mapping',
                    Value: 'DataReversalMappingActions',
                    HasInternalAction: true
                },
                {
                    Index: 22,
                    Name: '保存数据后',
                    Value: 'AfterSaveActions',
                    HasInternalAction: false
                },
                {
                    Index: 23,
                    Name: '变更集反向Mapping',
                    Value: 'ChangesetReversalMappingActions',
                    HasInternalAction: true
                }
            ]
        }];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Action执行参数
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * Action执行参数
     * @abstract
     */
    var  /**
     * Action执行参数
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * Action执行参数
     * @abstract
     */
    ViewModelParameter = /** @class */ (function () {
        function ViewModelParameter() {
        }
        return ViewModelParameter;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var VMParameterMode = {
        IN: 'IN',
        OUT: 'OUT',
        INOUT: 'INOUT',
    };
    /** @type {?} */
    var VMParameterModes = [
        { text: 'IN', value: VMParameterMode.IN },
        { text: 'OUT', value: VMParameterMode.OUT },
        { text: 'INOUT', value: VMParameterMode.INOUT }
    ];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * ViewModel的返回值实体类
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * Action执行参数
     */
    var  /**
     * ViewModel的返回值实体类
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * Action执行参数
     */
    ViewModelReturnValue = /** @class */ (function (_super) {
        __extends(ViewModelReturnValue, _super);
        function ViewModelReturnValue() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.ParamName = undefined;
            _this.Mode = VMParameterMode.OUT;
            return _this;
        }
        return ViewModelReturnValue;
    }(ViewModelParameter));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ViewModelVoidReturnValue = /** @class */ (function (_super) {
        __extends(ViewModelVoidReturnValue, _super);
        function ViewModelVoidReturnValue() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.Assembly = 'mscorlib.dll';
            _this.ClassName = 'Void';
            return _this;
        }
        return ViewModelVoidReturnValue;
    }(ViewModelReturnValue));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * VM Action抽象类
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * VM Action抽象类
     * @abstract
     */
    var  /**
     * VM Action抽象类
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * VM Action抽象类
     * @abstract
     */
    ViewModelAction = /** @class */ (function () {
        function ViewModelAction() {
            // this.ID = Guid.newGuid();
            this.ReturnValue = new ViewModelVoidReturnValue();
        }
        return ViewModelAction;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * VO操作类型枚举
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /** @enum {string} */
    var ViewModelActionType = {
        /**
         * BE的业务操作
         */
        BEAction: 'BEAction',
        /**
         * VM 操作
         */
        VMAction: 'VMAction',
        /**
         * 自定义操作
         */
        Custom: 'Custom',
    };
    /** @type {?} */
    var ViewModelActionTypes = [
        { text: 'BE操作', value: ViewModelActionType.BEAction },
        { text: 'VM操作', value: ViewModelActionType.VMAction },
    ];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * VM上映射的BE操作类
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * VM上映射的BE操作类
     */
    var  /**
     * VM上映射的BE操作类
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * VM上映射的BE操作类
     */
    MappedBizAction = /** @class */ (function (_super) {
        __extends(MappedBizAction, _super);
        function MappedBizAction() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            /**
             * 类型
             */
            _this.Type = ViewModelActionType.BEAction;
            return _this;
        }
        return MappedBizAction;
    }(ViewModelAction));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * VO上映射的构件操作类
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * VO上映射的构件操作类
     */
    var  /**
     * VO上映射的构件操作类
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * VO上映射的构件操作类
     */
    MappedCdpAction = /** @class */ (function (_super) {
        __extends(MappedCdpAction, _super);
        function MappedCdpAction() {
            var _this = _super.call(this) || this;
            /**
             * 类型
             */
            _this.Type = ViewModelActionType.VMAction;
            _this.IsGenerateComponent = true;
            return _this;
        }
        return MappedCdpAction;
    }(ViewModelAction));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Action执行结果
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * Action执行结果
     */
    var /**
     * Action执行结果
     */ VMActionResult = /** @class */ (function () {
        function VMActionResult() {
        }
        return VMActionResult;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 模板VO信息
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 模板VO信息
     */
    var /**
     * 模板VO信息
     */ TemplateVoInfo = /** @class */ (function () {
        function TemplateVoInfo() {
        }
        return TemplateVoInfo;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Action执行参数
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * Action执行参数
     */
    var  /**
     * Action执行参数
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * Action执行参数
     */
    MappedBizActionParameter = /** @class */ (function (_super) {
        __extends(MappedBizActionParameter, _super);
        function MappedBizActionParameter() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return MappedBizActionParameter;
    }(ViewModelParameter));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Action执行参数
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * Action执行参数
     */
    var  /**
     * Action执行参数
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * Action执行参数
     */
    MappedCdpActionParameter = /** @class */ (function (_super) {
        __extends(MappedCdpActionParameter, _super);
        function MappedCdpActionParameter() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return MappedCdpActionParameter;
    }(ViewModelParameter));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var VMParameterType = {
        /**
         * 字符型
         */
        String: 'String',
        /**
         * 布尔型
         */
        Boolean: 'Boolean',
        /**
         * 整数
         */
        Int32: 'Int32',
        /**
         * 浮点数
         */
        Decimal: 'Decimal',
        /**
         * 双浮点
         */
        Double: 'Double',
        /**
         * 时间
         */
        DateTime: 'DateTime',
        /**
         * 对象类型
         */
        Object: 'Object',
        /**
         * 自定义
         */
        Custom: 'Custom',
    };
    /** @type {?} */
    var VMParameterTypes = [
        { index: 0, text: '字符串', value: VMParameterType.String, assembly: 'mscorlib.dll', className: 'System.String', javaClassName: 'java.lang.String' },
        { index: 1, text: '布尔型', value: VMParameterType.Boolean, assembly: 'mscorlib.dll', className: 'System.Boolean', javaClassName: 'java.lang.Boolean' },
        { index: 2, text: '整数', value: VMParameterType.Int32, assembly: 'mscorlib.dll', className: 'System.Int32', javaClassName: 'java.lang.Integer' },
        { index: 3, text: '浮点数', value: VMParameterType.Decimal, assembly: 'mscorlib.dll', className: 'System.Single', javaClassName: 'java.math.BigDecimal' },
        { index: 4, text: '双浮点数', value: VMParameterType.Double, assembly: 'mscorlib.dll', className: 'System.Double', javaClassName: 'java.lang.Double' },
        { index: 5, text: '时间类型', value: VMParameterType.DateTime, assembly: 'mscorlib.dll', className: 'System.DateTime', javaClassName: 'java.util.Date' },
        { index: 6, text: '对象类型', value: VMParameterType.Object, assembly: 'mscorlib.dll', className: 'System.Object', javaClassName: 'java.lang.Object' },
        { index: 7, text: '自定义类型', value: VMParameterType.Custom, assembly: '', className: '', javaClassName: '' },
    ];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var VMCollectionParameterType = {
        /**
         * 未使用集合形式
         */
        None: 'None',
        /**
         * 列表
         */
        List: 'List',
        /**
         * 数组
         */
        Array: 'Array',
    };
    /** @type {?} */
    var VMCollectionParameterTypes = [
        { index: 0, text: '无', value: VMCollectionParameterType.None },
        { index: 1, text: '列表', value: VMCollectionParameterType.List },
        { index: 2, text: '数组', value: VMCollectionParameterType.Array }
    ];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 值帮助配置类
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 值帮助配置类
     */
    var  /**
     * 值帮助配置类
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * 值帮助配置类
     */
    ValueHelpConfig = /** @class */ (function () {
        function ValueHelpConfig() {
        }
        return ValueHelpConfig;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * VO映射接口
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * VO映射接口
     */
    var  /**
     * VO映射接口
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * VO映射接口
     */
    ViewModelMapping = /** @class */ (function () {
        function ViewModelMapping() {
        }
        return ViewModelMapping;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 帮助扩展动作（帮助前，帮助后等等一系列帮助的扩展）
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /**
     * 帮助扩展动作（帮助前，帮助后等等一系列帮助的扩展）
     */
    var  /**
     * 帮助扩展动作（帮助前，帮助后等等一系列帮助的扩展）
     * @author wangjiegj<wangjiegj@inspur.com>
     */ 
    /**
     * 帮助扩展动作（帮助前，帮助后等等一系列帮助的扩展）
     */
    HelpExtendAction = /** @class */ (function () {
        function HelpExtendAction() {
        }
        return HelpExtendAction;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * VO映射接口
     */
    var /**
     * VO映射接口
     */ GspVoElementMapping = /** @class */ (function (_super) {
        __extends(GspVoElementMapping, _super);
        function GspVoElementMapping() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return GspVoElementMapping;
    }(ViewModelMapping));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * VO映射接口
     */
    var /**
     * VO映射接口
     */ GspVoObjectMapping = /** @class */ (function (_super) {
        __extends(GspVoObjectMapping, _super);
        function GspVoObjectMapping() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return GspVoObjectMapping;
    }(ViewModelMapping));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 权限控制类型
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /** @enum {string} */
    var ElementHelpType = {
        /**
         * 使用传递的权限，不限制权限，Action没有绑定的业务操作
         */
        TransAuth: 'TransAuth',
        /**
         * 使用传递的权限，并且限制在Action所绑定的业务操作内
         */
        TransAndInsideAuth: 'TransAndInsideAuth',
        /**
         * 不使用传递的权限，使用Action绑定的业务操作（只绑定一个业务操作）
         */
        MustHasAuth: 'MustHasAuth',
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 映射类型枚举
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /** @enum {string} */
    var MappingType = {
        /**
         * BE
         */
        BizEntity: 'BizEntity',
        /**
         * DM
         */
        DataModel: 'DataModel',
        /**
         * BE节点
         */
        BEObject: 'BEObject',
        /**
         * GspDataModel节点
         */
        DMObject: 'DMObject',
        /**
         * 节点元素
         */
        Element: 'Element',
        /**
         * BE操作
         */
        BizOperation: 'BizOperation',
        /**
         * 构件方法
         */
        ComponentMethod: 'ComponentMethod',
        /**
         * 数据对象
         */
        DataObject: 'DataObject',
        /**
         * 数据对象列
         */
        GSPColumn: 'GSPColumn',
    };
    /** @type {?} */
    var MappingTypes = [
        { text: '业务实体', value: MappingType.BizEntity },
        { text: '业务实体节点', value: MappingType.BEObject },
        { text: '业务实体字段', value: MappingType.Element },
        { text: '业务实体操作', value: MappingType.BizOperation },
        { text: '构件方法', value: MappingType.ComponentMethod },
    ];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 映射元数据类型枚举
     * @author wangjiegj<wangjiegj@inspur.com>
     */
    /** @enum {string} */
    var MetadataType = {
        /**
         * BE
         */
        BizEntity: 'BizEntity',
        /**
         * QO
         */
        QueryObject: 'QueryObject',
    };
    /** @type {?} */
    var MetadataTypes = [
        { text: '业务实体', value: MetadataType.BizEntity },
        { text: '查询', value: MetadataType.QueryObject },
    ];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var GspVoObjectSourceType = {
        BeObject: 'BeObject',
        QoObject: 'QoObject',
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var GspVoElementSourceType = {
        BeElement: 'BeElement',
        QoElement: 'QoElement',
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.ɵb = gspCmMetadata.GSPCommonElement;
    exports.ɵa = gspCmMetadata.GSPCommonModel;
    exports.ɵc = gspCmMetadata.GSPCommonObject;
    exports.GSPViewModel = GSPViewModel;
    exports.GSPViewModelElement = GSPViewModelElement;
    exports.GSPViewObject = GSPViewObject;
    exports.VoDataExtendInfo = VoDataExtendInfo;
    exports.VoDataExtendInfos = VoDataExtendInfos;
    exports.MappedBizAction = MappedBizAction;
    exports.MappedCdpAction = MappedCdpAction;
    exports.ViewModelAction = ViewModelAction;
    exports.VMActionResult = VMActionResult;
    exports.TemplateVoInfo = TemplateVoInfo;
    exports.ViewModelParameter = ViewModelParameter;
    exports.MappedBizActionParameter = MappedBizActionParameter;
    exports.MappedCdpActionParameter = MappedCdpActionParameter;
    exports.VMParameterMode = VMParameterMode;
    exports.VMParameterModes = VMParameterModes;
    exports.VMParameterType = VMParameterType;
    exports.VMParameterTypes = VMParameterTypes;
    exports.VMCollectionParameterType = VMCollectionParameterType;
    exports.VMCollectionParameterTypes = VMCollectionParameterTypes;
    exports.ViewModelReturnValue = ViewModelReturnValue;
    exports.ViewModelVoidReturnValue = ViewModelVoidReturnValue;
    exports.ValueHelpConfig = ValueHelpConfig;
    exports.ViewModelMapping = ViewModelMapping;
    exports.HelpExtendAction = HelpExtendAction;
    exports.GspVoElementMapping = GspVoElementMapping;
    exports.GspVoObjectMapping = GspVoObjectMapping;
    exports.ElementHelpType = ElementHelpType;
    exports.MappingType = MappingType;
    exports.MappingTypes = MappingTypes;
    exports.MetadataType = MetadataType;
    exports.MetadataTypes = MetadataTypes;
    exports.ViewModelActionType = ViewModelActionType;
    exports.ViewModelActionTypes = ViewModelActionTypes;
    exports.GspVoObjectSourceType = GspVoObjectSourceType;
    exports.GspVoElementSourceType = GspVoElementSourceType;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
