// @edp-pmf/bpmn-model/dist/cjs/lib/model/participant/PalParticipant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PalParticipant = void 0;
var tslib_1 = require("tslib");
var ExtensionMethod_1 = require("../ExtensionMethod");
var OrganizationLevel_1 = require("./OrganizationLevel");
var PalParticipantType_1 = require("./PalParticipantType");
var Participant_1 = require("./Participant");
var PalParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(PalParticipant, _super);
    function PalParticipant(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "pal.PositionParticipant";
        _this.palParticipantType = PalParticipantType_1.PalParticipantType.Position;
        // 组织
        _this.organizationLevel = OrganizationLevel_1.OrganizationLevel.DEPARTMENT;
        _this.positionOrgType = "Sys";
        return _this;
    }
    PalParticipant.prototype.NewInstance = function () {
        return new PalParticipant(this.Model);
    };
    PalParticipant.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.palParticipantType = PalParticipantType_1.PalParticipantType.Position;
        this.organizationLevel = OrganizationLevel_1.OrganizationLevel.DEPARTMENT;
        this.organizationId = null;
        this.organizationCode = null;
        this.organizationName = null;
        this.positionId = null;
        this.positionCode = null;
        this.positionName = null;
        this.positionType = null;
        this.positionOrgId = null;
        this.positionOrgName = null;
        this.positionOrgType = null;
        this.userId = null;
        this.userCode = null;
        this.userName = null;
    };
    PalParticipant.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "PalParticipantType", this.palParticipantType);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "OrganizationLevel", this.organizationLevel);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "OrganizationId", this.organizationId);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "OrganizationCode", this.organizationCode);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "OrganizationName", this.organizationName);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "PositionId", this.positionId);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "PositionCode", this.positionCode);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "PositionName", this.positionName);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "PositionType", this.positionType);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "PositionOrgId", this.positionOrgId);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "PositionOrgName", this.positionOrgName);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "PositionOrgType", this.positionOrgType);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "UserId", this.userId);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "UserCode", this.userCode);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "UserName", this.userName);
        return obj;
    };
    PalParticipant.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.palParticipantType = PalParticipantType_1.PalParticipantType[ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "PalParticipantType")];
        this.organizationLevel = OrganizationLevel_1.OrganizationLevel[ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "OrganizationLevel")];
        this.organizationId = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "OrganizationId");
        this.organizationCode = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "OrganizationCode");
        this.organizationName = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "OrganizationName");
        this.positionId = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "PositionId");
        this.positionCode = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "PositionCode");
        this.positionName = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "PositionName");
        this.positionType = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "PositionType");
        this.positionOrgId = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "PositionOrgId");
        this.positionOrgName = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "PositionOrgName");
        this.positionOrgType = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "PositionOrgType");
        this.userId = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "UserId");
        this.userCode = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "UserCode");
        this.userName = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "UserName");
    };
    PalParticipant.prototype.clone = function () {
        var c = this.NewInstance();
        c.palParticipantType = this.palParticipantType;
        c.organizationId = this.organizationId;
        c.organizationCode = this.organizationCode;
        c.organizationName = this.organizationName;
        c.organizationLevel = this.organizationLevel;
        c.positionId = this.positionId;
        c.positionCode = this.positionCode;
        c.positionName = this.positionName;
        c.positionOrgId = this.positionOrgId;
        c.positionOrgName = this.positionOrgName;
        c.positionOrgType = this.positionOrgType;
        c.positionType = this.positionType;
        c.userId = this.userId;
        c.userCode = this.userCode;
        c.userName = this.userName;
        return c;
    };
    return PalParticipant;
}(Participant_1.Participant));
exports.PalParticipant = PalParticipant;
