// @edp-pmf/bpmn-model/dist/cjs/lib/model/participant/Participant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Participant = void 0;
var tslib_1 = require("tslib");
var ExtensionElement_1 = require("../ExtensionElement");
var ExtensionMethod_1 = require("../ExtensionMethod");
var Participant = /** @class */ (function (_super) {
    tslib_1.__extends(Participant, _super);
    function Participant(model) {
        return _super.call(this, model) || this;
    }
    Participant.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "Id", this.id);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "Name", this.name);
        return obj;
    };
    Participant.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.id = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "Id");
        this.name = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "Name");
    };
    Participant.prototype.clone = function () {
        var c = this.NewInstance();
        c.id = this.id;
        c.name = this.name;
        return c;
    };
    return Participant;
}(ExtensionElement_1.ExtensionElement));
exports.Participant = Participant;
