// @farris/ui-datatable/esm2015/lib/utils/drag.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, NgZone, Renderer2 } from '@angular/core';
import { DataTableComponent } from '../datatable.component';
export class DragColumnDirective {
    /**
     * @param {?} dt
     * @param {?} ngzone
     * @param {?} el
     * @param {?} render
     */
    constructor(dt, ngzone, el, render) {
        this.dt = dt;
        this.ngzone = ngzone;
        this.el = el;
        this.render = render;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        // if (this.isEnable()) {
        //     this.ngzone.runOutsideAngular(() => {
        //         // this.resizer = document.createElement('span');
        //         // this.resizer.className = 'column-resizer';
        //         // this.el.nativeElement.appendChild(this.resizer);
        //         // this.render.addClass(this.el.nativeElement, 'resizable-column');
        //         this.resizer = this.el.nativeElement.querySelector('span.column-resizer');
        //         if (this.resizer) {
        //             this.resizerMouseDownListener = this.onMouseDown.bind(this);
        //             this.resizer.addEventListener('mousedown', this.resizerMouseDownListener);
        //         }
        //     });
        // }
    }
    /**
     * @return {?}
     */
    isEnable() {
        return this.dt.resizableColumns && !this.dt.hasFixed;
    }
    /**
     * @return {?}
     */
    bindDocumentEvents() {
        this.ngzone.runOutsideAngular((/**
         * @return {?}
         */
        () => {
            this.documentMouseMoveListener = this.onMouseMove.bind(this);
            document.addEventListener('mousemove', this.documentMouseMoveListener);
            this.documentMouseUpListener = this.onMouseUp.bind(this);
            document.addEventListener('mouseup', this.documentMouseUpListener);
        }));
    }
    /**
     * @return {?}
     */
    unbindDocumentEvents() {
        if (this.documentMouseMoveListener) {
            document.removeEventListener('mousemove', this.documentMouseMoveListener);
            this.documentMouseMoveListener = null;
        }
        if (this.documentMouseUpListener) {
            document.removeEventListener('mouseup', this.documentMouseUpListener);
            this.documentMouseUpListener = null;
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onMouseDown(event) {
        this.dt.beginDrag(event);
        this.bindDocumentEvents();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onMouseMove(event) {
        this.dt.moveDrag(event);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onMouseUp(event) {
        this.dt.stopDrag(event, this.el.nativeElement);
        this.unbindDocumentEvents();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        // if (this.resizerMouseDownListener && this.resizer) {
        //     this.resizer.removeEventListener('mousedown', this.resizerMouseDownListener);
        // }
        this.unbindDocumentEvents();
    }
}
DragColumnDirective.decorators = [
    { type: Directive, args: [{
                selector: '[drag-column]',
                exportAs: 'resizeRef'
            },] }
];
/** @nocollapse */
DragColumnDirective.ctorParameters = () => [
    { type: DataTableComponent },
    { type: NgZone },
    { type: ElementRef },
    { type: Renderer2 }
];
if (false) {
    /** @type {?} */
    DragColumnDirective.prototype.resizer;
    /** @type {?} */
    DragColumnDirective.prototype.resizerMouseDownListener;
    /** @type {?} */
    DragColumnDirective.prototype.documentMouseMoveListener;
    /** @type {?} */
    DragColumnDirective.prototype.documentMouseUpListener;
    /** @type {?} */
    DragColumnDirective.prototype.dt;
    /** @type {?} */
    DragColumnDirective.prototype.ngzone;
    /** @type {?} */
    DragColumnDirective.prototype.el;
    /** @type {?} */
    DragColumnDirective.prototype.render;
}
