// @farris/devkit/esm5/lib/listener/ui_state_change_listener.js
import * as tslib_1 from "tslib";
import { Injectable, Injector, Inject } from '@angular/core';
// import { FRAME_ID, NAMESPACE } from '../frame/index';
import { UIState } from '../ui-state/index';
import { ChangeListener } from './change_listener';
import { Expression } from '../expression/index';
import { BindingData } from '../binding-data/index';
import { FRAME_ID, NAMESPACE } from '../frame/tokens';
/**
 * 监听UIState变更
 */
var UIStateChangeListener = /** @class */ (function (_super) {
    tslib_1.__extends(UIStateChangeListener, _super);
    function UIStateChangeListener(injector, uiState, namespace, frameId, bindingData) {
        var _this = _super.call(this) || this;
        _this.injector = injector;
        _this.uiState = uiState;
        _this.namespace = namespace;
        _this.frameId = frameId;
        _this.bindingData = bindingData;
        _this.registerEvent();
        return _this;
    }
    UIStateChangeListener.prototype.buildEventPath = function (change) {
        return null;
    };
    UIStateChangeListener.prototype.registerEvent = function () {
        var _this = this;
        if (this.uiState && this.uiState.changes) {
            this.uiState.changes.subscribe(function (change) {
                var modification = {
                    ns: _this.namespace,
                    path: [change.field],
                    type: Expression.EventType.ValueChanged,
                    value: change.value,
                    source: Expression.EventSource.State,
                    frameId: _this.frameId
                };
                // console.log("UIStateChangeListener", modification);
                _this.subject.next(modification);
            });
        }
    };
    UIStateChangeListener.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    UIStateChangeListener.ctorParameters = function () { return [
        { type: Injector },
        { type: UIState },
        { type: undefined, decorators: [{ type: Inject, args: [NAMESPACE,] }] },
        { type: String, decorators: [{ type: Inject, args: [FRAME_ID,] }] },
        { type: BindingData }
    ]; };
    return UIStateChangeListener;
}(ChangeListener));
export { UIStateChangeListener };
