// @edp-aif/common-api/esm5/lib/JsonConvertableImpl.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
var /**
 * @abstract
 */
JsonConvertableImpl = /** @class */ (function () {
    function JsonConvertableImpl() {
    }
    /**
     * @return {?}
     */
    JsonConvertableImpl.prototype.ConvertToJson = /**
     * @return {?}
     */
    function () {
        return JSON.stringify(this.ConvertToJsonObject());
    };
    /**
     * @param {?} json
     * @return {?}
     */
    JsonConvertableImpl.prototype.LoadFromJson = /**
     * @param {?} json
     * @return {?}
     */
    function (json) {
        this.ResetBeforeLoadFrom();
        /** @type {?} */
        var obj = JSON.parse(json);
        this.LoadFromJsonObject(obj);
    };
    /**
     * @return {?}
     */
    JsonConvertableImpl.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = {};
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    JsonConvertableImpl.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        this.ResetBeforeLoadFrom();
    };
    /**从 JsonObject 中加载数据前，赋予各属性初始值。*/
    /**
     * 从 JsonObject 中加载数据前，赋予各属性初始值。
     * @protected
     * @return {?}
     */
    JsonConvertableImpl.prototype.ResetBeforeLoadFrom = /**
     * 从 JsonObject 中加载数据前，赋予各属性初始值。
     * @protected
     * @return {?}
     */
    function () {
    };
    return JsonConvertableImpl;
}());
/**
 * @abstract
 */
export { JsonConvertableImpl };
