// @farris/bef/esm2015/public-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export { ChangeDetailType } from './lib/types';
export { BE_SERVER_URI_TOKEN, BE_SESSION_HANDLING_STRATEGY_TOKEN } from './lib/tokens';
export { BefChangeBuilder } from './lib/bef_change_builder';
export { BefChangeHandler } from './lib/bef_change_handler';
export { BefSessionService } from './lib/session/index';
export { BefRestService } from './lib/bef_rest_service';
export { BefProxy } from './lib/bef_proxy';
export { BefLookupRestService } from './lib/bef_lookup.service';
export { BefLookupDefaultService } from './lib/bef_lookup_default.service';
export { BefRepository } from './lib/bef_repository';
export { FrameworkSessionService } from './lib/framework_session_service';
export { NG_VARIABLE, NgVariable } from './lib/decorators';
export { BefRepositoryUtil } from './lib/bef_repository_util';
export { BefDataPathUtil } from './lib/bef_data_path_util';
export { UriService } from './lib/uri_service';
export { Uri } from './lib/uri';
export { DefaultComboHttpService } from './lib/bef_combo.service';
export { DefaultDynamiControlGroupHttpService } from './lib/bef_dynamic_control_group_default.service';
export { BefEntityManager } from './lib/bef_entity_manager';
export { DynamicBefRepository } from './lib/dynamic_bef_repository';
export { PresetBefRepository } from './lib/preset_bef_repository';
export { BefVariableManager } from './lib/bef_variable_manager';
/** @type {?} */
export const VERSION = 'ver.';
