// @farris/ui-datagrid-filter/esm5/lib/filter-editors/filter-row-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var FilterRowEditorComponent = /** @class */ (function () {
    function FilterRowEditorComponent() {
        this.id = '';
    }
    /**
     * @return {?}
     */
    FilterRowEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.id = "filter-row_" + this.column.field;
    };
    /**
     * @return {?}
     */
    FilterRowEditorComponent.prototype.clearValue = /**
     * @return {?}
     */
    function () { };
    FilterRowEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'filter-editor',
                    template: ''
                }] }
    ];
    /** @nocollapse */
    FilterRowEditorComponent.ctorParameters = function () { return []; };
    FilterRowEditorComponent.propDecorators = {
        column: [{ type: Input }]
    };
    return FilterRowEditorComponent;
}());
export { FilterRowEditorComponent };
if (false) {
    /** @type {?} */
    FilterRowEditorComponent.prototype.column;
    /** @type {?} */
    FilterRowEditorComponent.prototype.id;
}
