// @farris/ui-pagination/esm5/lib/pagination.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { IdService } from '@farris/ui-common';
import { EventEmitter, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var PaginationService = /** @class */ (function () {
    function PaginationService() {
        this.change = new EventEmitter();
        this.instances = {};
        this.DEFAULT_ID = 'DEFAULT_PAGINATION_ID';
    }
    /**
     * @return {?}
     */
    PaginationService.prototype.defaultId = /**
     * @return {?}
     */
    function () {
        return this.DEFAULT_ID + new IdService().uuid();
    };
    /**
     * @param {?} id
     * @return {?}
     */
    PaginationService.prototype.removeInstance = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        if (this.instances[id]) {
            this.instances[id] = null;
            delete this.instances[id];
        }
    };
    /**
     * @param {?} instance
     * @return {?}
     */
    PaginationService.prototype.register = /**
     * @param {?} instance
     * @return {?}
     */
    function (instance) {
        if (instance.id == null) {
            instance.id = this.DEFAULT_ID;
        }
        if (!this.instances[instance.id]) {
            this.instances[instance.id] = instance;
            this.change.emit(instance.id);
        }
        else {
            /** @type {?} */
            var changed = this.updateInstance(instance);
            if (changed) {
                this.change.emit(instance.id);
            }
        }
    };
    /**
     * Check each property of the instance and update any that have changed. Return
     * true if any changes were made, else return false.
     */
    /**
     * Check each property of the instance and update any that have changed. Return
     * true if any changes were made, else return false.
     * @private
     * @param {?} instance
     * @return {?}
     */
    PaginationService.prototype.updateInstance = /**
     * Check each property of the instance and update any that have changed. Return
     * true if any changes were made, else return false.
     * @private
     * @param {?} instance
     * @return {?}
     */
    function (instance) {
        /** @type {?} */
        var changed = false;
        for (var prop in this.instances[instance.id]) {
            if (instance[prop] !== this.instances[instance.id][prop]) {
                this.instances[instance.id][prop] = instance[prop];
                changed = true;
            }
        }
        return changed;
    };
    /**
     * Returns the current page number.
     */
    /**
     * Returns the current page number.
     * @param {?} id
     * @return {?}
     */
    PaginationService.prototype.getCurrentPage = /**
     * Returns the current page number.
     * @param {?} id
     * @return {?}
     */
    function (id) {
        if (this.instances[id]) {
            return this.instances[id].currentPage;
        }
    };
    /**
     * Sets the current page number.
     */
    /**
     * Sets the current page number.
     * @param {?} id
     * @param {?} page
     * @return {?}
     */
    PaginationService.prototype.setCurrentPage = /**
     * Sets the current page number.
     * @param {?} id
     * @param {?} page
     * @return {?}
     */
    function (id, page) {
        if (this.instances[id]) {
            /** @type {?} */
            var instance = this.instances[id];
            /** @type {?} */
            var maxPage = Math.ceil(instance.totalItems / instance.itemsPerPage);
            if (page <= maxPage && 1 <= page) {
                this.instances[id].currentPage = page;
                this.change.emit(id);
            }
        }
    };
    /**
     * @param {?} id
     * @param {?} newList
     * @return {?}
     */
    PaginationService.prototype.setPageList = /**
     * @param {?} id
     * @param {?} newList
     * @return {?}
     */
    function (id, newList) {
        if (this.instances[id]) {
            /** @type {?} */
            var instance = this.instances[id];
            instance.pageList = newList;
            this.change.emit(id);
        }
    };
    /**
     * Sets the value of instance.totalItems
     */
    /**
     * Sets the value of instance.totalItems
     * @param {?} id
     * @param {?} totalItems
     * @return {?}
     */
    PaginationService.prototype.setTotalItems = /**
     * Sets the value of instance.totalItems
     * @param {?} id
     * @param {?} totalItems
     * @return {?}
     */
    function (id, totalItems) {
        if (this.instances[id] && 0 <= totalItems) {
            this.instances[id].totalItems = totalItems;
            this.change.emit(id);
        }
    };
    /**
     * Sets the value of instance.itemsPerPage.
     */
    /**
     * Sets the value of instance.itemsPerPage.
     * @param {?} id
     * @param {?} itemsPerPage
     * @return {?}
     */
    PaginationService.prototype.setItemsPerPage = /**
     * Sets the value of instance.itemsPerPage.
     * @param {?} id
     * @param {?} itemsPerPage
     * @return {?}
     */
    function (id, itemsPerPage) {
        if (this.instances[id]) {
            this.instances[id].itemsPerPage = itemsPerPage;
            this.change.emit(id);
        }
    };
    /**
     * Returns a clone of the pagination instance object matching the id. If no
     * id specified, returns the instance corresponding to the default id.
     */
    /**
     * Returns a clone of the pagination instance object matching the id. If no
     * id specified, returns the instance corresponding to the default id.
     * @param {?=} id
     * @return {?}
     */
    PaginationService.prototype.getInstance = /**
     * Returns a clone of the pagination instance object matching the id. If no
     * id specified, returns the instance corresponding to the default id.
     * @param {?=} id
     * @return {?}
     */
    function (id) {
        if (id === void 0) { id = this.DEFAULT_ID; }
        if (this.instances[id]) {
            return this.clone(this.instances[id]);
        }
        return (/** @type {?} */ ({}));
    };
    /**
     * Perform a shallow clone of an object.
     */
    /**
     * Perform a shallow clone of an object.
     * @private
     * @param {?} obj
     * @return {?}
     */
    PaginationService.prototype.clone = /**
     * Perform a shallow clone of an object.
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var target = {};
        for (var i in obj) {
            if (obj.hasOwnProperty(i)) {
                target[i] = obj[i];
            }
        }
        return target;
    };
    PaginationService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ PaginationService.ngInjectableDef = i0.defineInjectable({ factory: function PaginationService_Factory() { return new PaginationService(); }, token: PaginationService, providedIn: "root" });
    return PaginationService;
}());
export { PaginationService };
if (false) {
    /** @type {?} */
    PaginationService.prototype.change;
    /**
     * @type {?}
     * @private
     */
    PaginationService.prototype.instances;
    /**
     * @type {?}
     * @private
     */
    PaginationService.prototype.DEFAULT_ID;
}
