// @farris/ui-pagination/esm5/lib/pagination-controls.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PaginationControlsDirective } from './pagination-controls.directive';
import { Component, Input, Output, EventEmitter, ChangeDetectionStrategy, ViewEncapsulation, ViewChild, ElementRef } from '@angular/core';
/**
 * @param {?} input
 * @return {?}
 */
function coerceToBoolean(input) {
    return !!input && input !== 'false';
}
// styleUrls: ['./pagination-controls.component.css'],
/**
 * The default pagination controls component. Actually just a default implementation of a custom template.
 */
//  #farris-gotopagenumber{ text-align: center;
//     -moz-appearance: textfield;
//     width: 40px;
// }
// #farris-gotopagenumber::-webkit-outer-spin-button,
// #farris-gotopagenumber::-webkit-inner-spin-button {
//     -webkit-appearance: none;
// }
var PaginationControlsComponent = /** @class */ (function () {
    function PaginationControlsComponent() {
        this._directionLinks = true;
        this._autoHide = false;
        this._responsive = false;
        this.mouseInSelectedList = false;
        /**
         * 显示模式， default 默认；simple: 简洁
         */
        this.viewMode = 'default';
        this.maxSize = 7;
        this.previousLabel = 'Previous';
        this.nextLabel = 'Next';
        this.screenReaderPaginationLabel = 'Pagination';
        this.screenReaderPageLabel = 'page';
        this.screenReaderCurrentLabel = "You're on page";
        this.showPageList = true;
        this.showPageInfo = true;
        this.showPageNumber = true;
        this.showSelectedList = false;
        this.showGoToLast = false;
        this.showGoToFirst = false;
        /**
         * 显示跳转至指定页码输入框
         */
        this.showGotoInput = false;
        this.position = 'right';
        /**
         * @deprecated
         */
        this.message = '每页 {0} 条记录，共 {1} 条记录';
        /**
         * @deprecated
         */
        this.pageListFirstText = '显示';
        /**
         * @deprecated
         */
        this.pageListLastText = '条';
        this.pageChange = new EventEmitter();
        this.pageSizeChange = new EventEmitter();
        this.showPagesPanel = false;
    }
    Object.defineProperty(PaginationControlsComponent.prototype, "directionLinks", {
        get: /**
         * @return {?}
         */
        function () {
            return this._directionLinks;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._directionLinks = coerceToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PaginationControlsComponent.prototype, "autoHide", {
        get: /**
         * @return {?}
         */
        function () {
            return this._autoHide;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._autoHide = coerceToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PaginationControlsComponent.prototype, "responsive", {
        get: /**
         * @return {?}
         */
        function () {
            return this._responsive;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._responsive = coerceToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    // constructor(@Optional() private pagerService: PaginationService) {
    //     if (!this.pagerService) {
    //         this.pagerService = new PaginationService();
    //     }
    // }
    // constructor(@Optional() private pagerService: PaginationService) {
    //     if (!this.pagerService) {
    //         this.pagerService = new PaginationService();
    //     }
    // }
    /**
     * @param {?} $event
     * @return {?}
     */
    PaginationControlsComponent.prototype.onMouseLeave = 
    // constructor(@Optional() private pagerService: PaginationService) {
    //     if (!this.pagerService) {
    //         this.pagerService = new PaginationService();
    //     }
    // }
    /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.showPagesPanel = false;
        this.paginationDirective.changeDetectorRef.detectChanges();
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    PaginationControlsComponent.prototype.onMouseEnter = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.paginationDirective.getTotalItems() === 0) {
            return;
        }
        this.showPagesPanel = true;
    };
    /**
     * @param {?} $event
     * @param {?} pagesize
     * @return {?}
     */
    PaginationControlsComponent.prototype.changePageSizeHandler = /**
     * @param {?} $event
     * @param {?} pagesize
     * @return {?}
     */
    function ($event, pagesize) {
        $event.stopPropagation();
        // 当前页再次点击
        if (pagesize === this.paginationDirective.getPageSize()) {
            return;
        }
        this.showPagesPanel = false;
        // this.removeOverlay();
        this.pageSizeChange.emit(pagesize);
    };
    /**
     * @param {?} inputEl
     * @param {?} maxNumber
     * @return {?}
     */
    PaginationControlsComponent.prototype.onPageNumberChanged = /**
     * @param {?} inputEl
     * @param {?} maxNumber
     * @return {?}
     */
    function (inputEl, maxNumber) {
        /** @type {?} */
        var target = (/** @type {?} */ (inputEl.target));
        /** @type {?} */
        var currentPageNumber = Number(target.value);
        if (currentPageNumber > maxNumber) {
            target.value = maxNumber;
        }
        else {
            // if (currentPageNumber < 1) {
            //     target.value = 1;
            // }
        }
    };
    /**
     * @param {?} inputEl
     * @return {?}
     */
    PaginationControlsComponent.prototype.onPageNumberBlur = /**
     * @param {?} inputEl
     * @return {?}
     */
    function (inputEl) {
        /** @type {?} */
        var target = (/** @type {?} */ (inputEl.target));
        /** @type {?} */
        var currentPageNumber = Number(target.value);
        if (currentPageNumber < 1) {
            target.value = 1;
        }
    };
    /**
     * @param {?} newpageList
     * @return {?}
     */
    PaginationControlsComponent.prototype.setPageList = /**
     * @param {?} newpageList
     * @return {?}
     */
    function (newpageList) {
        this.paginationDirective.setPageList(newpageList);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    PaginationControlsComponent.prototype.goto = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var newpager = Number($event.target.value);
        if (newpager < 1) {
            newpager = 1;
            $event.target.value = 1;
        }
        if (this.paginationDirective.getCurrent() !== newpager) {
            this.paginationDirective.goto(newpager);
        }
    };
    PaginationControlsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'pagination-controls',
                    template: "<pagination-template #p=\"paginationApi\" [id]=\"id\" [maxSize]=\"maxSize\" (pageChange)=\"pageChange.emit($event)\">\r\n    <div class=\"pagination-container\">\r\n        <ng-content select=\".pager-overlay-container\"></ng-content>\r\n        <ul class=\"ngx-pagination pagination\" role=\"navigation\" [attr.aria-label]=\"screenReaderPaginationLabel\"\r\n            style=\"position: relative;align-items: center;\" [class.responsive]=\"responsive\"\r\n            [class.pager-viewmode-default]=\"viewMode === 'default'\"\r\n            [class.pager-viewmode-simple]=\"viewMode === 'simple'\" *ngIf=\"!(autoHide && p.pages.length <= 1)\"\r\n            [style.justify-content]=\"position === 'center'?'center': 'start'\">\r\n            <ng-container *ngIf=\"viewMode === 'default'\">\r\n                <li class=\"pagination-message text-truncate d-flex\" *ngIf=\"showPageInfo\" style=\"min-width: 70px\"\r\n                    [class.ml-auto]=\"position=== 'right'\" [class.flex-fill]=\"position=== 'right'\"\r\n                    [style.justify-content]=\"position === 'right' ? 'right': ''\">\r\n                    <ng-content select=\".pager-pos-right\"></ng-content>\r\n                    <ng-content></ng-content>\r\n                    <div class=\"text-truncate \">\r\n                        <span class=\"pg-message-text\">{{ 'pagination.totalinfo.firstText' | locale}}</span>\r\n                        <b class=\"pg-message-total\">{{ p.getTotalItems()}}</b>\r\n                        <span class=\"pg-message-text\">{{ 'pagination.totalinfo.lastText' | locale}}</span>\r\n                    </div>\r\n                </li>\r\n                <!--\u6BCF\u9875\u8BB0\u5F55\u6570-->\r\n                <li class=\"pagination-pagelist\" *ngIf=\"showPageList\">\r\n                    <div class=\"dropup dropdown-right pg-pagelist\"\r\n                        [ngClass]=\"{'pagelist-disabled':p.getTotalItems()===0,'show': showPagesPanel}\"\r\n                        (mouseenter)=\"onMouseEnter($event)\" (mouseleave)=\"onMouseLeave($event)\">\r\n                        <div class=\"pg-pagelist-info\" #pglistinfo>\r\n                            <span class=\"pagelist-text\">{{'pagination.pagelist.firstText' | locale}}</span><b\r\n                                class=\"cur-pagesize\">{{p.getPageSize()}}</b><span\r\n                                class=\"pagelist-text\">{{'pagination.pagelist.lastText' | locale}}</span><i\r\n                                class=\"f-icon f-icon-dropdown\"></i>\r\n                        </div>\r\n                        <div class=\"dropdown-menu\" [ngClass]=\"{'show': showPagesPanel}\" style=\"margin-bottom: -1px;\">\r\n                <li class=\"dropdown-item\" [ngClass]=\"{'active':pagesize==p.getPageSize()}\"\r\n                    *ngFor=\"let pagesize of p.getPageList()\" (click)=\"changePageSizeHandler($event, pagesize)\">\r\n                    <span>{{ pagesize }}</span>\r\n                </li>\r\n    </div>\r\n    </div>\r\n    </li>\r\n\r\n    <!--\u9996\u9875-->\r\n    <li class=\"page-item\" [class.disabled]=\"p.isFirstPage()\" *ngIf=\"directionLinks\">\r\n        <a tabindex=\"0\" *ngIf=\"1 < p.getCurrent()\" (keyup.enter)=\"p.first($event)\" (click)=\"p.first($event)\"\r\n            [attr.aria-label]=\"previousLabel + ' ' + screenReaderPageLabel\" class=\"page-link\">\r\n            <span class=\"f-icon f-page-first\"></span>\r\n        </a>\r\n        <span *ngIf=\"p.isFirstPage()\" class=\"page-link\">\r\n            <span class=\"f-icon f-page-first\"></span>\r\n        </span>\r\n    </li>\r\n\r\n    <!--\u4E0A\u4E00\u9875-->\r\n    <li class=\"page-item\" [class.disabled]=\"p.isFirstPage()\" *ngIf=\"directionLinks\">\r\n        <a class=\"page-link\" tabindex=\"0\" *ngIf=\"1 < p.getCurrent()\" (keyup.enter)=\"p.previous()\" (click)=\"p.previous()\"\r\n            [attr.aria-label]=\"previousLabel + ' ' + screenReaderPageLabel\">\r\n            <span class=\"f-icon f-page-pre\"></span>\r\n        </a>\r\n        <span *ngIf=\"p.isFirstPage()\" class=\"page-link\">\r\n            <span class=\"f-icon f-page-pre\"></span>\r\n        </span>\r\n    </li>\r\n\r\n\r\n\r\n    <ng-container *ngIf=\"showPageNumber\">\r\n        <!--\u9875\u7801-->\r\n        <li class=\"page-item\" [class.current]=\"p.getCurrent() === page.value\" [class.ellipsis]=\"page.label === '...'\"\r\n            *ngFor=\"let page of p.pages\">\r\n            <a class=\"page-link\" tabindex=\"0\" (keyup.enter)=\"p.setCurrent(page.value)\"\r\n                (click)=\"p.setCurrent(page.value)\" *ngIf=\"p.getCurrent() !== page.value\">\r\n                <span>{{ page.label }}</span>\r\n            </a>\r\n            <span class=\"page-link\" *ngIf=\"p.getCurrent() === page.value\">{{page.label }}</span>\r\n        </li>\r\n    </ng-container>\r\n    <!--\u4E0B\u4E00\u9875-->\r\n    <li class=\"page-item\" [class.disabled]=\"p.isLastPage()\" *ngIf=\"directionLinks\">\r\n        <a class=\"page-link\" tabindex=\"0\" *ngIf=\"!p.isLastPage()\" (keyup.enter)=\"p.next()\" (click)=\"p.next()\"\r\n            [attr.aria-label]=\"nextLabel + ' ' + screenReaderPageLabel\">\r\n            <span class=\"f-icon f-page-next\"></span>\r\n        </a>\r\n        <span *ngIf=\"p.isLastPage()\" class=\"page-link\">\r\n            <span class=\"f-icon f-page-next\"></span>\r\n        </span>\r\n    </li>\r\n    <!--\u672B\u9875-->\r\n    <li class=\"page-item\" [class.disabled]=\"p.isLastPage()\" *ngIf=\"directionLinks\">\r\n        <a class=\"page-link\" tabindex=\"0\" *ngIf=\"!p.isLastPage()\" (keyup.enter)=\"p.last($event)\"\r\n            (click)=\"p.last($event)\" [attr.aria-label]=\"nextLabel + ' ' + screenReaderPageLabel\">\r\n            <span class=\"f-icon f-page-last\"></span>\r\n        </a>\r\n        <span *ngIf=\"p.isLastPage()\" class=\"page-link\">\r\n            <span class=\"f-icon f-page-last\"></span>\r\n        </span>\r\n    </li>\r\n\r\n\r\n\r\n    <li class=\"page-goto-input d-flex flex-row\" *ngIf=\"showGotoInput\"\r\n        style=\"padding-left: 10px; white-space: nowrap;align-items: center;\">\r\n        {{'pagination.goto.prefix' | locale}}\r\n\r\n        <div style=\"min-width: 40px;position: relative;height: 24px;margin: 0 3px;\">\r\n            <label style=\"display: inline-block;font-size: inherit;line-height: normal;visibility: hidden;font-family: inherit;\r\n            padding: 0;\">{{gotoinput.value}}</label>\r\n            <input #gotoinput class=\"form-control farris-gotopagenumber\" value=\"1\" type=\"number\" min=\"1\"\r\n                max=\"{{p.getLastPage()}}\"\r\n                style=\"width: 100%;position: absolute;left: 0;top: 0;padding: 0;height: 100%;line-height: 24px;\"\r\n                (input)=\"onPageNumberChanged($event, p.getLastPage())\" [value]=\"p.getCurrent()\"\r\n                (keyup.enter)=\"goto($event)\" (blur)=\"onPageNumberBlur($event)\">\r\n        </div>\r\n\r\n\r\n        {{'pagination.goto.suffix' | locale}}\r\n    </li>\r\n\r\n    <li class=\"pagination-message text-truncate d-flex flex-fill ml-3\" *ngIf=\"showPageInfo && position==='left'\">\r\n        <ng-content select=\".pager-pos-left\"></ng-content>\r\n    </li>\r\n\r\n    </ng-container>\r\n    <ng-container *ngIf=\"viewMode === 'simple'\">\r\n        <li class=\"page-item d-flex flex-fill\">\r\n            <ng-content select=\".simple-pager-extend-area\"></ng-content>\r\n        </li>\r\n        <!--\u9996\u9875-->\r\n        <li class=\"page-item\" [class.disabled]=\"p.isFirstPage()\" *ngIf=\"directionLinks && showGoToFirst\">\r\n            <a tabindex=\"0\" *ngIf=\"1 < p.getCurrent()\" (keyup.enter)=\"p.first($event)\" (click)=\"p.first($event)\"\r\n                [attr.aria-label]=\"previousLabel + ' ' + screenReaderPageLabel\" class=\"page-link\">\r\n                <span class=\"f-icon f-page-first\"></span>\r\n            </a>\r\n            <span *ngIf=\"p.isFirstPage()\" class=\"page-link\">\r\n                <span class=\"f-icon f-page-first\"></span>\r\n            </span>\r\n        </li>\r\n        <!--\u4E0A\u4E00\u9875-->\r\n        <li class=\"page-item\" [class.disabled]=\"p.isFirstPage()\" *ngIf=\"directionLinks\">\r\n            <a class=\"page-link border-0\" tabindex=\"0\" *ngIf=\"1 < p.getCurrent()\" (keyup.enter)=\"p.previous()\"\r\n                (click)=\"p.previous()\" [attr.aria-label]=\"previousLabel + ' ' + screenReaderPageLabel\">\r\n                <span class=\"f-icon f-page-pre\"></span>\r\n            </a>\r\n            <span *ngIf=\"p.isFirstPage()\" class=\"page-link border-0\">\r\n                <span class=\"f-icon f-page-pre\"></span>\r\n            </span>\r\n        </li>\r\n\r\n        <li class=\"page-item\">\r\n            <input id=\"farris-gotopagenumber\" class=\"form-control farris-gotopagenumber\" min=\"1\"\r\n                max=\"{{p.getLastPage()}}\" type=\"number\" [ngModel]=\"p.getCurrent()\"\r\n                (input)=\"onPageNumberChanged($event, p.getLastPage())\" (keyup.enter)=\"p.goto($event.target.value)\">\r\n        </li>\r\n        <li class=\"page-item page-separator\" style=\"margin-left: 10px\">\r\n            <span style=\"font-size: 15px; font-weight: 200;\"> /</span>\r\n        </li>\r\n        <li class=\"page-item page-total\" style=\"margin-left: 5px\">\r\n            <span style=\"font-size: 16px; font-weight: 600;\"> {{ p.getLastPage() }}</span>\r\n        </li>\r\n\r\n        <!--\u4E0B\u4E00\u9875-->\r\n        <li class=\"page-item\" [class.disabled]=\"p.isLastPage()\" *ngIf=\"directionLinks\">\r\n            <a class=\"page-link border-0\" tabindex=\"0\" *ngIf=\"!p.isLastPage()\" (keyup.enter)=\"p.next()\"\r\n                (click)=\"p.next()\" [attr.aria-label]=\"nextLabel + ' ' + screenReaderPageLabel\">\r\n                <span class=\"f-icon f-page-next\"></span>\r\n            </a>\r\n            <span *ngIf=\"p.isLastPage()\" class=\"page-link border-0\">\r\n                <span class=\"f-icon f-page-next\"></span>\r\n            </span>\r\n        </li>\r\n        <!--\u672B\u9875-->\r\n        <li class=\"page-item\" [class.disabled]=\"p.isLastPage()\" *ngIf=\"directionLinks  && showGoToLast\">\r\n            <a class=\"page-link\" tabindex=\"0\" *ngIf=\"!p.isLastPage()\" (keyup.enter)=\"p.last($event)\"\r\n                (click)=\"p.last($event)\" [attr.aria-label]=\"nextLabel + ' ' + screenReaderPageLabel\">\r\n                <span class=\"f-icon f-page-last\"></span>\r\n            </a>\r\n            <span *ngIf=\"p.isLastPage()\" class=\"page-link\">\r\n                <span class=\"f-icon f-page-last\"></span>\r\n            </span>\r\n        </li>\r\n    </ng-container>\r\n\r\n    </ul>\r\n    </div>\r\n</pagination-template>",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    PaginationControlsComponent.propDecorators = {
        viewMode: [{ type: Input }],
        id: [{ type: Input }],
        maxSize: [{ type: Input }],
        directionLinks: [{ type: Input }],
        autoHide: [{ type: Input }],
        responsive: [{ type: Input }],
        previousLabel: [{ type: Input }],
        nextLabel: [{ type: Input }],
        screenReaderPaginationLabel: [{ type: Input }],
        screenReaderPageLabel: [{ type: Input }],
        screenReaderCurrentLabel: [{ type: Input }],
        showPageList: [{ type: Input }],
        showPageInfo: [{ type: Input }],
        showPageNumber: [{ type: Input }],
        showSelectedList: [{ type: Input }],
        showGoToLast: [{ type: Input }],
        showGoToFirst: [{ type: Input }],
        showGotoInput: [{ type: Input }],
        position: [{ type: Input }],
        message: [{ type: Input }],
        pageListFirstText: [{ type: Input }],
        pageListLastText: [{ type: Input }],
        pageChange: [{ type: Output }],
        pageSizeChange: [{ type: Output }],
        paginationDirective: [{ type: ViewChild, args: ['p',] }],
        focusPageSizeRef: [{ type: ViewChild, args: ['focusPageSize',] }],
        pglistinfo: [{ type: ViewChild, args: ['pglistinfo',] }]
    };
    return PaginationControlsComponent;
}());
export { PaginationControlsComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PaginationControlsComponent.prototype._directionLinks;
    /**
     * @type {?}
     * @private
     */
    PaginationControlsComponent.prototype._autoHide;
    /**
     * @type {?}
     * @private
     */
    PaginationControlsComponent.prototype._responsive;
    /**
     * @type {?}
     * @private
     */
    PaginationControlsComponent.prototype.mouseInSelectedList;
    /**
     * 显示模式， default 默认；simple: 简洁
     * @type {?}
     */
    PaginationControlsComponent.prototype.viewMode;
    /** @type {?} */
    PaginationControlsComponent.prototype.id;
    /** @type {?} */
    PaginationControlsComponent.prototype.maxSize;
    /** @type {?} */
    PaginationControlsComponent.prototype.previousLabel;
    /** @type {?} */
    PaginationControlsComponent.prototype.nextLabel;
    /** @type {?} */
    PaginationControlsComponent.prototype.screenReaderPaginationLabel;
    /** @type {?} */
    PaginationControlsComponent.prototype.screenReaderPageLabel;
    /** @type {?} */
    PaginationControlsComponent.prototype.screenReaderCurrentLabel;
    /** @type {?} */
    PaginationControlsComponent.prototype.showPageList;
    /** @type {?} */
    PaginationControlsComponent.prototype.showPageInfo;
    /** @type {?} */
    PaginationControlsComponent.prototype.showPageNumber;
    /** @type {?} */
    PaginationControlsComponent.prototype.showSelectedList;
    /** @type {?} */
    PaginationControlsComponent.prototype.showGoToLast;
    /** @type {?} */
    PaginationControlsComponent.prototype.showGoToFirst;
    /**
     * 显示跳转至指定页码输入框
     * @type {?}
     */
    PaginationControlsComponent.prototype.showGotoInput;
    /** @type {?} */
    PaginationControlsComponent.prototype.position;
    /**
     * @deprecated
     * @type {?}
     */
    PaginationControlsComponent.prototype.message;
    /**
     * @deprecated
     * @type {?}
     */
    PaginationControlsComponent.prototype.pageListFirstText;
    /**
     * @deprecated
     * @type {?}
     */
    PaginationControlsComponent.prototype.pageListLastText;
    /** @type {?} */
    PaginationControlsComponent.prototype.pageChange;
    /** @type {?} */
    PaginationControlsComponent.prototype.pageSizeChange;
    /** @type {?} */
    PaginationControlsComponent.prototype.paginationDirective;
    /** @type {?} */
    PaginationControlsComponent.prototype.focusPageSizeRef;
    /** @type {?} */
    PaginationControlsComponent.prototype.pglistinfo;
    /** @type {?} */
    PaginationControlsComponent.prototype.showPagesPanel;
}
