// @gsp-svc/expression/esm2015/lib/expression-schema-info-grid/expression-schema-info-grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { Property } from '../schema';
import { CurrentNode } from '../currentNode';
export class ExpressionSchemaInfoGridComponent {
    constructor() {
        this.schema = new Map();
        this.choseSchemaTableName = new CurrentNode;
        this.selectSchemaRow = new EventEmitter();
        this.dbclickInsert = new EventEmitter();
        this.selectString = "";
        this.items = [];
        this.columns = [{ field: 'name', title: '属性名称', filter: true }];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngOnChanges() {
        /** @type {?} */
        let schemaName = this.choseSchemaTableName.parentName;
        /** @type {?} */
        let tableId = this.choseSchemaTableName.id;
        /** @type {?} */
        let tableName = this.choseSchemaTableName.name;
        /** @type {?} */
        let schemaItem = this.schema.get(schemaName);
        if (schemaItem) {
            /** @type {?} */
            let entityType = schemaItem.entityTypes.find((/**
             * @param {?} entitytype
             * @return {?}
             */
            entitytype => (entitytype.id == tableId && entitytype.name == tableName) || (entitytype.id == "" && entitytype.name == tableName)));
            if (entityType) {
                this.items = entityType.properties;
                entityType.navigationProperties.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => {
                    /** @type {?} */
                    let property = new Property;
                    property.name = item.name;
                    property.type = item.type.toString();
                    property.description = item.description;
                    this.items = this.items.filter((/**
                     * @param {?} item
                     * @return {?}
                     */
                    item => item.name !== property.name));
                    this.items.push(property);
                }));
                this.selectString = entityType.name;
            }
            else {
                /** @type {?} */
                let entitySet = schemaItem.entitySets.find((/**
                 * @param {?} entityset
                 * @return {?}
                 */
                entityset => entityset.name == tableName));
                if (entitySet) {
                    /** @type {?} */
                    let eType = schemaItem.entityTypes.find((/**
                     * @param {?} entitytype
                     * @return {?}
                     */
                    entitytype => entitytype.name == entitySet.entityType));
                    this.items = eType.properties;
                    eType.navigationProperties.forEach((/**
                     * @param {?} item
                     * @return {?}
                     */
                    item => {
                        /** @type {?} */
                        let property = new Property;
                        property.name = item.name;
                        property.type = item.type.toString();
                        property.description = item.description;
                        this.items = this.items.filter((/**
                         * @param {?} item
                         * @return {?}
                         */
                        item => item.name !== property.name));
                        this.items.push(property);
                    }));
                    this.selectString = entitySet.name + ".[0]";
                }
            }
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    ondblclickRow($event) {
        this.dbclickInsert.emit(this.selectString + "." + $event.rowData.name);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    selectRowChange($event) {
        this.selectSchemaRow.emit($event.data);
    }
}
ExpressionSchemaInfoGridComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-expression-schema-info-grid',
                template: "<farris-datagrid\r\n[idField]=\"'name'\"\r\n[height]=\"268\"\r\n[width]=\"300\"\r\n[columns]=\"columns\" \r\n[data]=\"items\"\r\n[fit]=\"true\"\r\n[fitColumns]=\"true\"\r\n[striped]=\"false\"\r\n[showHeader]=\"true\"\r\n[pagination]=\"false\"\r\n[enableFilterRow]=\"true\"\r\n(dblclickRow)=\"ondblclickRow($event)\"\r\n(selectChanged)=\"selectRowChange($event)\"\r\n>\r\n</farris-datagrid>",
                styles: [""]
            }] }
];
/** @nocollapse */
ExpressionSchemaInfoGridComponent.ctorParameters = () => [];
ExpressionSchemaInfoGridComponent.propDecorators = {
    schema: [{ type: Input }],
    choseSchemaTableName: [{ type: Input }],
    selectSchemaRow: [{ type: Output }],
    dbclickInsert: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    ExpressionSchemaInfoGridComponent.prototype.schema;
    /** @type {?} */
    ExpressionSchemaInfoGridComponent.prototype.choseSchemaTableName;
    /** @type {?} */
    ExpressionSchemaInfoGridComponent.prototype.selectSchemaRow;
    /** @type {?} */
    ExpressionSchemaInfoGridComponent.prototype.dbclickInsert;
    /** @type {?} */
    ExpressionSchemaInfoGridComponent.prototype.selectString;
    /** @type {?} */
    ExpressionSchemaInfoGridComponent.prototype.items;
    /** @type {?} */
    ExpressionSchemaInfoGridComponent.prototype.columns;
}
