// @gsp-wf/ui-comment/esm5/lib/task-comment/task-comment.viewmodel.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { TaskCommentUIState } from './task-comment.uistate';
import { CommonComment } from '../entities/comment';
import { forkJoin } from 'rxjs';
import { HttpService } from '@ecp-caf/caf-common';
var TaskCommentViewModel = /** @class */ (function () {
    function TaskCommentViewModel(uiState, httpSvc) {
        this.uiState = uiState;
        this.httpSvc = httpSvc;
    }
    /**
     * @return {?}
     */
    TaskCommentViewModel.prototype.add = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var newEntity = CommonComment.newEntity();
        newEntity.sortOrder = this.uiState.commentList.length + 1;
        this.uiState.commentList.push(newEntity);
    };
    /**
     * @return {?}
     */
    TaskCommentViewModel.prototype.delete = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.uiState.curComment) {
            /** @type {?} */
            var index = this.uiState.commentList.findIndex(function (c) { return c.id === _this.uiState.curComment.id; });
            if (index > -1) {
                this.uiState.commentList.splice(index, 1);
            }
        }
    };
    /**
     * @return {?}
     */
    TaskCommentViewModel.prototype.save = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var url = "/api/runtime/wf/v1.0/commonComments";
        /** @type {?} */
        var deleteIds = this.uiState.originCommentIds.filter(function (id) { return _this.uiState.commentList.findIndex(function (c) { return c.id === id; }) < 0; });
        /** @type {?} */
        var addEntities = this.uiState.commentList.filter(function (item) { return _this.uiState.originCommentIds.indexOf(item.id) < 0; });
        /** @type {?} */
        var changesList = this.uiState.commentList.filter(function (item) { return addEntities.findIndex(function (aItem) { return aItem.id === item.id; }) < 0
            && _this.uiState.changesCommentIds.indexOf(item.id) > -1; });
        return forkJoin(deleteIds.map(function (id) { return _this.httpSvc.delete(url + ("/" + id), null); })
            .concat(changesList.map(function (change) { return _this.httpSvc.patch(url, change); }))
            .concat(addEntities.map(function (add) { return _this.httpSvc.post(url, add); })));
    };
    TaskCommentViewModel.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    TaskCommentViewModel.ctorParameters = function () { return [
        { type: TaskCommentUIState },
        { type: HttpService }
    ]; };
    return TaskCommentViewModel;
}());
export { TaskCommentViewModel };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaskCommentViewModel.prototype.uiState;
    /**
     * @type {?}
     * @private
     */
    TaskCommentViewModel.prototype.httpSvc;
}
