// @farris/mobile-ui/lib/staff-card/index.js
import { renderSlot, createVNode, toDisplayString, createTextVNode } from 'vue';

var script = {
  name: 'fm-staff-card',
  props: {
    staffName: {
      type: String
    },
    title: {
      type: String
    },
    departmentName: {
      type: String
    },
    staffImage: {
      type: String
    }
  }
};

var _hoisted_1 = {
  "class": "fm-staff-card"
};
var _hoisted_2 = {
  "class": "fm-staff-card-image"
};
var _hoisted_3 = {
  "class": "fm-staff-card-main"
};
var _hoisted_4 = {
  "class": "staff-main-text"
};

var _hoisted_5 = /*#__PURE__*/createTextVNode("的 ");

var _hoisted_6 = {
  "class": "pl-base"
};
var _hoisted_7 = {
  "class": "staff-main-subtext"
};
function render(_ctx, _cache, $props, $setup, $data, $options) {
  return renderSlot(_ctx.$slots, "default", {}, function () {
    return [createVNode("div", _hoisted_1, [createVNode("div", _hoisted_2, [createVNode("img", {
      src: $props.staffImage,
      alt: "",
      "class": "fm-staff-card-img"
    }, null, 8
    /* PROPS */
    , ["src"])]), createVNode("div", _hoisted_3, [createVNode("div", _hoisted_4, [createVNode("span", null, toDisplayString($props.staffName), 1
    /* TEXT */
    ), _hoisted_5, createVNode("span", _hoisted_6, "【" + toDisplayString($props.title) + "】", 1
    /* TEXT */
    )]), createVNode("div", _hoisted_7, toDisplayString($props.departmentName), 1
    /* TEXT */
    )])])];
  });
}

script.render = render;
script.__file = "packages/staff-card/src/index.vue";

script.install = function (app) {
  app.component(script.name, script);
};

export default script;
export { script as StaffCard };
