// @gsp-cmp/ccmp-ui/esm5/lib/ccmp-selector/common-component-selector.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output, ViewChild, TemplateRef, ComponentFactoryResolver, Injector } from '@angular/core';
import { ViewUtils } from '../view-utils/view-utils';
import { ComponentInfo } from '../params/component-info';
import { BsModalService } from '@farris/ui-modal';
import { ComponentConvertor } from '../convertor/convertor';
var CommonComponentSelectorComponent = /** @class */ (function () {
    function CommonComponentSelectorComponent(utils, modalService, resolver, injector) {
        var _this = this;
        this.utils = utils;
        this.modalService = modalService;
        this.resolver = resolver;
        this.injector = injector;
        this.selectedOpUris = [];
        this.isRowSelected = (/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return _this.selectedOpUris.indexOf(e.dataItem.uri) >= 0; });
        this.confirmEmitter = new EventEmitter();
        this.closeEmitter = new EventEmitter();
    }
    Object.defineProperty(CommonComponentSelectorComponent.prototype, "componentInfo", {
        set: /**
         * @param {?} _componentInfo
         * @return {?}
         */
        function (_componentInfo) {
            this._componentInfo = _componentInfo;
            this.metadata = _componentInfo.metadata;
            /** @type {?} */
            var convertor = new ComponentConvertor();
            this.metadata = convertor.initFromObject(convertor.convertorToObject(_componentInfo.metadata));
            this.metadata.operations = new Array();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    CommonComponentSelectorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._componentInfo = this._componentInfo || new ComponentInfo();
    };
    /**
     * @return {?}
     */
    CommonComponentSelectorComponent.prototype.confirm = /**
     * @return {?}
     */
    function () {
        if (!this._componentInfo.metadata || !this._componentInfo.metadata.operations) {
            if (this._componentInfo.metadata.operations.length <= 0) {
                this.utils.showInfo('没有可选的构件操作', 'info', '', 1500);
                return;
            }
        }
        if (!this.metadata.operations || this.metadata.operations.length <= 0) {
            this.utils.showInfo('请选择构件操作', 'info', '', 1500);
            return;
        }
        if (this.dialog) {
            this.dialog.close();
        }
        this.confirmEmitter.emit(this.metadata);
    };
    /**
     * @return {?}
     */
    CommonComponentSelectorComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        if (this.dialog) {
            this.dialog.close();
        }
        this.closeEmitter.emit();
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    CommonComponentSelectorComponent.prototype.cellClickHandler = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.selectedOpUris = [$event.dataItem.uri];
        this.metadata.operations = [$event.dataItem];
    };
    CommonComponentSelectorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'common-component-selector',
                    template: "<div class=\"d-flex\" style=\"position: absolute;top: 0;bottom: 0;left: 0;right: 0;min-width: 300px;\">\n    <kendo-grid class=\"f-utils-flex-column\" style=\"flex:1 1 auto;padding: 0px 8px 0px 8px;\" [data]=\"_componentInfo?.metadata?.operations\"\n        [selectable]=\"{enabled: true, checkboxOnly: false,mode:'single' }\" [kendoGridSelectBy]=\"'uri'\" (cellClick)=\"cellClickHandler($event)\"\n        [rowSelected]=\"isRowSelected\">\n        <kendo-grid-column field=\"code\" title=\"\u64CD\u4F5C\u7F16\u53F7\"></kendo-grid-column>\n        <kendo-grid-column field=\"name\" title=\"\u64CD\u4F5C\u540D\u79F0\"></kendo-grid-column>\n        <!-- <kendo-grid-column width=\"100\" title=\"\u53C2\u6570\u4FE1\u606F\">\n            <ng-template kendoGridCellTemplate let-dataItem>\n                <span style=\"float:left;font-size: 14px;color: #1890ff;cursor: pointer;\" (click)=\"view(dataItem)\">\u67E5\u770B<span\n                        class=\"f-icon f-accordion-collapse\"></span></span>\n            </ng-template>\n        </kendo-grid-column> -->\n    </kendo-grid>\n</div>\n\n<ng-template #btnRef>\n    <div style=\"width: 100%;line-height: 50px; text-align: center;min-width: 140px;\">\n        <button class=\"btn btn-secondary\" (click)=\"close()\" style=\"float:right;margin-right: 5px\">\u53D6\u6D88</button>\n        <button class=\"btn btn-primary\" (click)=confirm() style=\"float:right;margin-right: 5px\">\u786E\u5B9A</button>\n    </div>\n</ng-template>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    CommonComponentSelectorComponent.ctorParameters = function () { return [
        { type: ViewUtils },
        { type: BsModalService },
        { type: ComponentFactoryResolver },
        { type: Injector }
    ]; };
    CommonComponentSelectorComponent.propDecorators = {
        componentInfo: [{ type: Input }],
        btnRef: [{ type: ViewChild, args: ['btnRef',] }],
        confirmEmitter: [{ type: Output }],
        closeEmitter: [{ type: Output }]
    };
    return CommonComponentSelectorComponent;
}());
export { CommonComponentSelectorComponent };
if (false) {
    /** @type {?} */
    CommonComponentSelectorComponent.prototype._componentInfo;
    /** @type {?} */
    CommonComponentSelectorComponent.prototype.dialog;
    /** @type {?} */
    CommonComponentSelectorComponent.prototype.metadata;
    /** @type {?} */
    CommonComponentSelectorComponent.prototype.selectedOpUris;
    /** @type {?} */
    CommonComponentSelectorComponent.prototype.isRowSelected;
    /** @type {?} */
    CommonComponentSelectorComponent.prototype.btnRef;
    /** @type {?} */
    CommonComponentSelectorComponent.prototype.confirmEmitter;
    /** @type {?} */
    CommonComponentSelectorComponent.prototype.closeEmitter;
    /**
     * @type {?}
     * @private
     */
    CommonComponentSelectorComponent.prototype.utils;
    /**
     * @type {?}
     * @private
     */
    CommonComponentSelectorComponent.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    CommonComponentSelectorComponent.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    CommonComponentSelectorComponent.prototype.injector;
}
