// @farris/ui-responsive/esm2015/lib/col/res-col.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, HostBinding } from '@angular/core';
export class FarrisColDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        this.span = 0;
        // 向右移动几个col
        this.offset = 0;
        this.nativeClass = '';
        this.nativeClass += this.el.nativeElement.classList.value;
    }
    /**
     * @return {?}
     */
    get ColClass() {
        return `${this.nativeClass} ${this.getSpanClass()}${this.getBreakpointClass()}${this.getOffsetClass()}${this.getOrderClass()}`;
    }
    /**
     * @return {?}
     */
    get isCol() {
        return this.col;
    }
    /**
     * @return {?}
     */
    get isColSm() {
        return this.colSm;
    }
    /**
     * @return {?}
     */
    get isColMd() {
        return this.colMd;
    }
    /**
     * @return {?}
     */
    get isColLg() {
        return this.colLg;
    }
    /**
     * @return {?}
     */
    get isColXl() {
        return this.colXl;
    }
    /* 获取span 类名 */
    /**
     * @return {?}
     */
    getSpanClass() {
        return this.span ? ` col-${this.span}` : ` `;
    }
    /* 获取响应式 类名  xs sm md lg xl 可以是数字，代表span;也可以是对象，内部包含span offset*/
    /**
     * @return {?}
     */
    getBreakpointClass() {
        /** @type {?} */
        const makeClass = (/**
         * @param {?} breakpoint
         * @return {?}
         */
        (breakpoint) => {
            /** @type {?} */
            const props = this[breakpoint] || {};
            // 获取对象中  span或者offset
            return Object.keys(props).map((/**
             * @param {?} prop
             * @return {?}
             */
            prop => prop === 'span' ?
                `col-${breakpoint}-${props[prop]}` : prop === 'offset' ? `offset-${breakpoint}-${props[prop]}` :
                `order-${breakpoint}-${props[prop]}`)).join(' ');
        });
        return ['sm', 'md', 'lg', 'xl'].reduce((/**
         * @param {?} pre
         * @param {?} current
         * @return {?}
         */
        (pre, current) => {
            return typeof this[current] === 'object' ? `${pre} ${makeClass(current)}` :
                typeof this[current] === 'number' ? `${pre} col-${current}-${this[current]}` : `${pre} `;
        }), '');
    }
    /* 获取offset */
    /**
     * @return {?}
     */
    getOffsetClass() {
        return this.offset ? ` offset-${this.offset}` : ` `;
    }
    // get order class
    /**
     * @return {?}
     */
    getOrderClass() {
        return this.order ? ` order-${this.order}` : ` `;
    }
}
FarrisColDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-col]'
            },] }
];
/** @nocollapse */
FarrisColDirective.ctorParameters = () => [
    { type: ElementRef }
];
FarrisColDirective.propDecorators = {
    col: [{ type: Input }],
    colSm: [{ type: Input, args: ['col-sm',] }],
    colMd: [{ type: Input, args: ['col-md',] }],
    colLg: [{ type: Input, args: ['col-lg',] }],
    colXl: [{ type: Input, args: ['col-xl',] }],
    span: [{ type: Input }],
    offset: [{ type: Input }],
    order: [{ type: Input }],
    xs: [{ type: Input }],
    sm: [{ type: Input }],
    md: [{ type: Input }],
    lg: [{ type: Input }],
    xl: [{ type: Input }],
    ColClass: [{ type: HostBinding, args: ['class',] }],
    isCol: [{ type: HostBinding, args: ['class.col',] }],
    isColSm: [{ type: HostBinding, args: ['class.col-sm',] }],
    isColMd: [{ type: HostBinding, args: ['class.col-md',] }],
    isColLg: [{ type: HostBinding, args: ['class.col-lg',] }],
    isColXl: [{ type: HostBinding, args: ['class.col-xl',] }]
};
if (false) {
    /** @type {?} */
    FarrisColDirective.prototype.col;
    /** @type {?} */
    FarrisColDirective.prototype.colSm;
    /** @type {?} */
    FarrisColDirective.prototype.colMd;
    /** @type {?} */
    FarrisColDirective.prototype.colLg;
    /** @type {?} */
    FarrisColDirective.prototype.colXl;
    /** @type {?} */
    FarrisColDirective.prototype.span;
    /** @type {?} */
    FarrisColDirective.prototype.offset;
    /** @type {?} */
    FarrisColDirective.prototype.order;
    /** @type {?} */
    FarrisColDirective.prototype.xs;
    /** @type {?} */
    FarrisColDirective.prototype.sm;
    /** @type {?} */
    FarrisColDirective.prototype.md;
    /** @type {?} */
    FarrisColDirective.prototype.lg;
    /** @type {?} */
    FarrisColDirective.prototype.xl;
    /** @type {?} */
    FarrisColDirective.prototype.nativeClass;
    /**
     * @type {?}
     * @private
     */
    FarrisColDirective.prototype.el;
}
