// @gsp-svc/formdoc-upload/esm5/lib/formdoc-upload.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClientModule } from '@angular/common/http';
import { LOCALE_ID, NgModule } from '@angular/core';
import { CacheService, HttpService, LanguageService, SessionService } from '@ecp-caf/caf-common';
import { FormdocUploadComponent } from './formdoc-upload.component';
var ɵ0 = localStorage.getItem('languageCode') || 'zh-CHS';
var FormdocUploadModule = /** @class */ (function () {
    function FormdocUploadModule() {
    }
    FormdocUploadModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [FormdocUploadComponent],
                    imports: [
                        HttpClientModule
                    ],
                    exports: [FormdocUploadComponent],
                    providers: [LanguageService, HttpService, CacheService, SessionService,
                        { provide: LOCALE_ID, useValue: ɵ0 },]
                },] }
    ];
    return FormdocUploadModule;
}());
export { FormdocUploadModule };
export { ɵ0 };
