// @gsp-svc/formdoc-upload/esm5/lib/uploaddialog.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, ComponentFactoryResolver, LOCALE_ID } from '@angular/core';
import { FileUploadComponent } from './upload/file-upload/file-upload.component';
import { BsModalService } from '@farris/ui-modal';
import { of } from 'rxjs';
import { LocalLangPipe } from './local.pipe';
var UploadDialogService = /** @class */ (function () {
    function UploadDialogService(modalService, componentFactoryResolver, injector) {
        this.modalService = modalService;
        this.componentFactoryResolver = componentFactoryResolver;
        this.injector = injector;
        this.fileInfoList = [];
        if (this.injector) {
            this.modalService = this.injector.get(BsModalService);
            this.localeid = localStorage.getItem("languageCode") || this.injector.get(LOCALE_ID);
            this.localpipe = new LocalLangPipe(this.localeid);
        }
    }
    /**
     * @param {?} formId
     * @param {?} rootId
     * @return {?}
     */
    UploadDialogService.prototype.showDialog = /**
     * @param {?} formId
     * @param {?} rootId
     * @return {?}
     */
    function (formId, rootId) {
        var _this = this;
        /** @type {?} */
        var flag = 0;
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) {
            /** @type {?} */
            var viewerFactory = _this.componentFactoryResolver.resolveComponentFactory(FileUploadComponent);
            /** @type {?} */
            var viewerRef = viewerFactory.create(_this.injector);
            viewerRef.instance.rootId = rootId;
            viewerRef.instance.formId = formId;
            _this.dlg = _this.modalService.show(viewerRef, {
                title: _this.localpipe.transform('uploadFile'),
                width: 950, height: 570,
                buttons: [
                    {
                        text: _this.localpipe.transform('save'), cls: 'k-button k-button-icontext k-primary', handle: (/**
                         * @return {?}
                         */
                        function () {
                            _this.dlg.content.upload().subscribe((/**
                             * @param {?} res
                             * @return {?}
                             */
                            function (res) {
                                if (res) {
                                    resolve(res);
                                    flag = 1;
                                    _this.dlg.close();
                                }
                            }));
                        })
                    },
                    {
                        text: _this.localpipe.transform('close'), cls: 'k-button k-button-icontext', handle: (/**
                         * @return {?}
                         */
                        function () {
                            _this.dlg.close();
                        })
                    }
                ],
                showButtons: true,
                showMaxButton: false,
                beforeClose: (/**
                 * @return {?}
                 */
                function () {
                    if (flag == 0)
                        _this.dlg.content.cancel();
                    return of(true);
                })
            });
        }));
    };
    /**
     * @param {?} formId
     * @param {?} rootId
     * @param {?=} oldIdList
     * @return {?}
     */
    UploadDialogService.prototype.uploadFile = /**
     * @param {?} formId
     * @param {?} rootId
     * @param {?=} oldIdList
     * @return {?}
     */
    function (formId, rootId, oldIdList) {
        if (oldIdList === void 0) { oldIdList = []; }
        return this.uploadFileWithLimit(formId, rootId, null, oldIdList);
    };
    /**
     * @param {?} formId
     * @param {?} rootId
     * @param {?} limit
     * @param {?=} oldIdList
     * @return {?}
     */
    UploadDialogService.prototype.uploadFileWithLimit = /**
     * @param {?} formId
     * @param {?} rootId
     * @param {?} limit
     * @param {?=} oldIdList
     * @return {?}
     */
    function (formId, rootId, limit, oldIdList) {
        var _this = this;
        if (oldIdList === void 0) { oldIdList = []; }
        /** @type {?} */
        var flag = 0;
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) {
            /** @type {?} */
            var uploadFactory = _this.componentFactoryResolver.resolveComponentFactory(FileUploadComponent);
            /** @type {?} */
            var uploadRef = uploadFactory.create(_this.injector);
            if (limit != null) {
                uploadRef.instance.fileCount = limit.fileCount;
                uploadRef.instance.securityInfo = limit.securityInfo;
                if (limit.fileType != null && limit.fileType != "")
                    uploadRef.instance.fileType = limit.fileType;
            }
            uploadRef.instance.rootId = rootId;
            uploadRef.instance.formId = formId;
            uploadRef.instance.oldIdList = oldIdList;
            _this.dlg = _this.modalService.show(uploadRef, {
                title: _this.localpipe.transform('uploadFile'),
                width: 950, height: 570,
                buttons: [
                    {
                        text: _this.localpipe.transform('save'), cls: 'k-button k-button-icontext k-primary', handle: (/**
                         * @return {?}
                         */
                        function () {
                            _this.dlg.content.upload().subscribe((/**
                             * @param {?} res
                             * @return {?}
                             */
                            function (res) {
                                if (res) {
                                    resolve(res);
                                    flag = 1;
                                    _this.dlg.close();
                                }
                            }));
                        })
                    },
                    {
                        text: _this.localpipe.transform('close'), cls: 'k-button k-button-icontext', handle: (/**
                         * @return {?}
                         */
                        function () {
                            _this.dlg.close();
                        })
                    }
                ],
                showButtons: true,
                showMaxButton: false,
                beforeClose: (/**
                 * @return {?}
                 */
                function () {
                    if (flag == 0)
                        _this.dlg.content.cancel();
                    return of(true);
                })
            });
        }));
    };
    UploadDialogService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    UploadDialogService.ctorParameters = function () { return [
        { type: BsModalService },
        { type: ComponentFactoryResolver },
        { type: Injector }
    ]; };
    return UploadDialogService;
}());
export { UploadDialogService };
if (false) {
    /** @type {?} */
    UploadDialogService.prototype.localeid;
    /** @type {?} */
    UploadDialogService.prototype.localpipe;
    /** @type {?} */
    UploadDialogService.prototype.dlg;
    /** @type {?} */
    UploadDialogService.prototype.fileInfoList;
    /**
     * @type {?}
     * @private
     */
    UploadDialogService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    UploadDialogService.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    UploadDialogService.prototype.injector;
}
