// @farris/devkit/esm5/lib/variable/frame_id_variable_parser.js
/**
 * session变量解析
 * @author Witt <jiwt@inspur.com>
 */
import { Injectable } from '@angular/core';
import { ParseUtil } from './parse_util';
var VARIABLE_MATCHER = /#{\S+?}/g;
/**
 * 数据变量解析
 */
var FrameIdVariableParser = /** @class */ (function () {
    function FrameIdVariableParser() {
    }
    /**
     * 解析变量
     * @param expression 形如：/#{root-component}/stateName
     * @param context 上下文
     */
    FrameIdVariableParser.prototype.parse = function (expression, context) {
        var _this = this;
        this.context = context;
        var variables = this.extractVariables(expression);
        if (!variables || variables.length < 1) {
            return expression;
        }
        variables.forEach(function (exp) {
            var frameId = _this.getVariableValue(exp);
            expression = expression.replace(VARIABLE_MATCHER, frameId);
        });
        return expression;
    };
    /**
     * 计算表达式的值
     * @param expression like #{root-component}
     */
    FrameIdVariableParser.prototype.getVariableValue = function (expression) {
        var relativeFrameId = expression.substring(2, expression.length - 1);
        return this.getFullFrameId(relativeFrameId);
    };
    FrameIdVariableParser.prototype.extractVariables = function (expression) {
        if (!expression) {
            return [];
        }
        return expression.match(VARIABLE_MATCHER);
    };
    /**
     * 获取补全的frameId
     * @param frameId frameId
     */
    FrameIdVariableParser.prototype.getFullFrameId = function (frameId) {
        var frameContext = ParseUtil.getFrameContext(this.context);
        var namespace = frameContext.namespace || '';
        var prefix = namespace ? namespace + "_" : '';
        return "" + prefix + frameId;
    };
    FrameIdVariableParser.decorators = [
        { type: Injectable }
    ];
    return FrameIdVariableParser;
}());
export { FrameIdVariableParser };
