// @farris/devkit/esm5/lib/variable/index.js
import { VARIABLE_PARSERS } from './variable_parser';
import { DataVariableParser } from './data_variable_parser';
import { UIStateVariableParser } from './ui_state_variable_parser';
import { StateMachineVariableParser } from './state_machine_variable_parser';
import { CommandVariableParser } from './command_variable_parser';
import { VariableParseService } from './variable_parse_service';
import { FrameIdVariableParser } from './frame_id_variable_parser';
import { FormStateVariableParser } from './form_state_variable_parser';
import { EventParamVariableParser } from './event_param_variable_parser';
// Providers
var VARIABLE_PROVIDERS = [
    { provide: VARIABLE_PARSERS, multi: true, useClass: FrameIdVariableParser },
    { provide: VARIABLE_PARSERS, multi: true, useClass: FormStateVariableParser },
    { provide: VARIABLE_PARSERS, multi: true, useClass: EventParamVariableParser },
    { provide: VARIABLE_PARSERS, multi: true, useClass: DataVariableParser },
    { provide: VARIABLE_PARSERS, multi: true, useClass: UIStateVariableParser },
    { provide: VARIABLE_PARSERS, multi: true, useClass: StateMachineVariableParser },
    { provide: VARIABLE_PARSERS, multi: true, useClass: CommandVariableParser },
    VariableParseService
];
export { VARIABLE_PARSERS, FrameIdVariableParser, DataVariableParser, UIStateVariableParser, CommandVariableParser, StateMachineVariableParser, VariableParseService, EventParamVariableParser, VARIABLE_PROVIDERS };
