// @farris/extend-file-upload/esm5/lib/components/preview-checkbox/preview-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/preview-checkbox/preview-checkbox.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, HostBinding, } from "@angular/core";
var PreviewCheckboxComponent = /** @class */ (function () {
    function PreviewCheckboxComponent() {
        this.cls1 = true;
        this.cls2 = true;
        this.cls3 = true;
        this.checked = false;
        this.disabled = false;
        this.showDragBar = false;
        this.checkedChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    PreviewCheckboxComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} event
     * @return {?}
     */
    PreviewCheckboxComponent.prototype.clickHandler = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.stopPropagation();
        if (!this.disabled) {
            this.checked = !this.checked;
            this.checkedChange.emit({ checked: this.checked, id: this.id });
        }
    };
    PreviewCheckboxComponent.decorators = [
        { type: Component, args: [{
                    selector: "f-file-preview-checkbox",
                    template: "\r\n<span class=\"f-icon f-icon-drag-vertical upfile-drag-handle\" *ngIf=\"showDragBar\"></span>\r\n\r\n<div class=\"custom-control custom-checkbox f-checkradio-single  m-0\" >\r\n  <input class=\"custom-control-input\" type=\"checkbox\"  [checked]=\"checked\" />\r\n  <label class=\"custom-control-label\" (click)=\"clickHandler($event)\" (mousedown)=\"$event.stopPropagation()\"></label>\r\n</div>"
                }] }
    ];
    /** @nocollapse */
    PreviewCheckboxComponent.ctorParameters = function () { return []; };
    PreviewCheckboxComponent.propDecorators = {
        cls1: [{ type: HostBinding, args: ["class.align-middle",] }],
        cls2: [{ type: HostBinding, args: ["class.d-inline-flex",] }],
        cls3: [{ type: HostBinding, args: ["class.preview-checkbox",] }],
        checked: [{ type: Input }],
        id: [{ type: Input }],
        disabled: [{ type: Input }],
        showDragBar: [{ type: Input }],
        checkedChange: [{ type: Output }]
    };
    return PreviewCheckboxComponent;
}());
export { PreviewCheckboxComponent };
if (false) {
    /** @type {?} */
    PreviewCheckboxComponent.prototype.cls1;
    /** @type {?} */
    PreviewCheckboxComponent.prototype.cls2;
    /** @type {?} */
    PreviewCheckboxComponent.prototype.cls3;
    /** @type {?} */
    PreviewCheckboxComponent.prototype.checked;
    /** @type {?} */
    PreviewCheckboxComponent.prototype.id;
    /** @type {?} */
    PreviewCheckboxComponent.prototype.disabled;
    /** @type {?} */
    PreviewCheckboxComponent.prototype.showDragBar;
    /** @type {?} */
    PreviewCheckboxComponent.prototype.checkedChange;
}
