// @farris/ui-common/esm5/types/filter.conditions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var Compare = {
    Equal: 0,
    NotEqual: 1,
    Greater: 2,
    GreaterOrEqual: 3,
    Less: 4,
    LessOrEqual: 5,
    Like: 6,
    LikeStartWith: 7,
    LikeEndWith: 8,
    NotLike: 9,
    NotLikeStartWith: 10,
    NotLikeEndWith: 11,
    Is: 12,
    IsNot: 13,
    In: 14,
    NotIn: 15,
};
export { Compare };
Compare[Compare.Equal] = 'Equal';
Compare[Compare.NotEqual] = 'NotEqual';
Compare[Compare.Greater] = 'Greater';
Compare[Compare.GreaterOrEqual] = 'GreaterOrEqual';
Compare[Compare.Less] = 'Less';
Compare[Compare.LessOrEqual] = 'LessOrEqual';
Compare[Compare.Like] = 'Like';
Compare[Compare.LikeStartWith] = 'LikeStartWith';
Compare[Compare.LikeEndWith] = 'LikeEndWith';
Compare[Compare.NotLike] = 'NotLike';
Compare[Compare.NotLikeStartWith] = 'NotLikeStartWith';
Compare[Compare.NotLikeEndWith] = 'NotLikeEndWith';
Compare[Compare.Is] = 'Is';
Compare[Compare.IsNot] = 'IsNot';
Compare[Compare.In] = 'In';
Compare[Compare.NotIn] = 'NotIn';
/** @enum {number} */
var FilterRelation = {
    Empty: 0,
    And: 1,
    Or: 2,
};
export { FilterRelation };
FilterRelation[FilterRelation.Empty] = 'Empty';
FilterRelation[FilterRelation.And] = 'And';
FilterRelation[FilterRelation.Or] = 'Or';
/**
 * @record
 */
export function FilterCondition() { }
if (false) {
    /** @type {?} */
    FilterCondition.prototype.filterField;
    /** @type {?} */
    FilterCondition.prototype.value;
    /** @type {?|undefined} */
    FilterCondition.prototype.lbracket;
    /** @type {?|undefined} */
    FilterCondition.prototype.rbracket;
    /** @type {?|undefined} */
    FilterCondition.prototype.relation;
    /** @type {?|undefined} */
    FilterCondition.prototype.compare;
    /** @type {?|undefined} */
    FilterCondition.prototype.expresstype;
}
/** @enum {?} */
var ExpressValueType = {
    Value: 0,
    Expression: 1,
    FrontExpress: 'frontExpress',
};
export { ExpressValueType };
ExpressValueType[ExpressValueType.Value] = 'Value';
ExpressValueType[ExpressValueType.Expression] = 'Expression';
/** @type {?} */
export var LeftBracket = [
    '', '(', '((', '((('
];
/** @type {?} */
export var RighttBracket = [
    '', ')', '))', ')))'
];
/** @type {?} */
export var Relation = [
    { label: '', value: 0 },
    { label: '并且', value: 1 },
    { label: '或者', value: 2 }
];
/** @type {?} */
export var CompareOperators = [
    { label: '=', value: 0 },
    { label: '<>', value: 1 },
    { label: '>', value: 2 },
    { label: '>=', value: 3 },
    { label: '<', value: 4 },
    { label: '<=', value: 5 },
    { label: 'like %...%', value: 6 },
    { label: 'like ...%', value: 7 },
    { label: 'like %...', value: 8 },
    { label: 'not like %...%', value: 9 },
    { label: 'not like ...%', value: 10 },
    { label: 'not like %...', value: 11 },
    { label: 'is', value: 12 },
    { label: 'is not', value: 13 },
    { label: 'in', value: 14 },
    { label: 'not in', value: 15 },
];
