// @gsp-wf/ui-comment/esm5/lib/services/task-comment.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
var TaskCommentService = /** @class */ (function () {
    function TaskCommentService(httpSvc) {
        this.httpSvc = httpSvc;
    }
    /**
     * @return {?}
     */
    TaskCommentService.newId = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var S4 = function () { return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1); };
        return (S4() + S4() + '-' + S4() + '-' + S4() + '-' + S4() + '-' + S4() + S4() + S4());
    };
    /**
     * @return {?}
     */
    TaskCommentService.prototype.getCommentList = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var url = "/api/runtime/wf/v1.0/commonComments";
        return this.httpSvc.get(url);
    };
    /**
     * @return {?}
     */
    TaskCommentService.prototype.addComment = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var url = "/api/runtime/wf/v1.0/commonComments";
        return this.httpSvc.post(url, {});
    };
    /**
      * 获取任务中心配置项
      * @param code 配置项的Code
      */
    /**
     * 获取任务中心配置项
     * @param {?} code 配置项的Code
     * @return {?}
     */
    TaskCommentService.prototype.getTaskCenterConfigByCode = /**
     * 获取任务中心配置项
     * @param {?} code 配置项的Code
     * @return {?}
     */
    function (code) {
        /** @type {?} */
        var url = "/api/runtime/task/v1.0/task-center-setting/code?code=" + code;
        return this.httpSvc.get(url);
    };
    /**
     * 真正上传附件
     * @param files
     */
    /**
     * 真正上传附件
     * @param {?} files
     * @return {?}
     */
    TaskCommentService.prototype.uploadAttachment = /**
     * 真正上传附件
     * @param {?} files
     * @return {?}
     */
    function (files) {
        /** @type {?} */
        var url = "/api/runtime/dfs/v1.0/doc/commit";
        return this.httpSvc.put(url, files);
    };
    /**
     * 删除附件
     * @param fileDeleteEntity
     */
    /**
     * 删除附件
     * @param {?} fileDeleteEntity
     * @return {?}
     */
    TaskCommentService.prototype.deleteAttachment = /**
     * 删除附件
     * @param {?} fileDeleteEntity
     * @return {?}
     */
    function (fileDeleteEntity) {
        /** @type {?} */
        var url = "/api/runtime/dfs/v1.0/doc";
        return this.httpSvc.delete(url, fileDeleteEntity);
    };
    TaskCommentService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    TaskCommentService.ctorParameters = function () { return [
        { type: HttpService }
    ]; };
    return TaskCommentService;
}());
export { TaskCommentService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaskCommentService.prototype.httpSvc;
}
