// @farris/ui-responsive/esm5/lib/col/res-col.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, HostBinding } from '@angular/core';
var FarrisColDirective = /** @class */ (function () {
    function FarrisColDirective(el) {
        this.el = el;
        this.span = 0;
        // 向右移动几个col
        this.offset = 0;
        this.nativeClass = '';
        this.nativeClass += this.el.nativeElement.classList.value;
    }
    Object.defineProperty(FarrisColDirective.prototype, "ColClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.nativeClass + " " + this.getSpanClass() + this.getBreakpointClass() + this.getOffsetClass() + this.getOrderClass();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FarrisColDirective.prototype, "isCol", {
        get: /**
         * @return {?}
         */
        function () {
            return this.col;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FarrisColDirective.prototype, "isColSm", {
        get: /**
         * @return {?}
         */
        function () {
            return this.colSm;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FarrisColDirective.prototype, "isColMd", {
        get: /**
         * @return {?}
         */
        function () {
            return this.colMd;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FarrisColDirective.prototype, "isColLg", {
        get: /**
         * @return {?}
         */
        function () {
            return this.colLg;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FarrisColDirective.prototype, "isColXl", {
        get: /**
         * @return {?}
         */
        function () {
            return this.colXl;
        },
        enumerable: true,
        configurable: true
    });
    /* 获取span 类名 */
    /* 获取span 类名 */
    /**
     * @return {?}
     */
    FarrisColDirective.prototype.getSpanClass = /* 获取span 类名 */
    /**
     * @return {?}
     */
    function () {
        return this.span ? " col-" + this.span : " ";
    };
    /* 获取响应式 类名  xs sm md lg xl 可以是数字，代表span;也可以是对象，内部包含span offset*/
    /* 获取响应式 类名  xs sm md lg xl 可以是数字，代表span;也可以是对象，内部包含span offset*/
    /**
     * @return {?}
     */
    FarrisColDirective.prototype.getBreakpointClass = /* 获取响应式 类名  xs sm md lg xl 可以是数字，代表span;也可以是对象，内部包含span offset*/
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var makeClass = (/**
         * @param {?} breakpoint
         * @return {?}
         */
        function (breakpoint) {
            /** @type {?} */
            var props = _this[breakpoint] || {};
            // 获取对象中  span或者offset
            return Object.keys(props).map((/**
             * @param {?} prop
             * @return {?}
             */
            function (prop) { return prop === 'span' ?
                "col-" + breakpoint + "-" + props[prop] : prop === 'offset' ? "offset-" + breakpoint + "-" + props[prop] :
                "order-" + breakpoint + "-" + props[prop]; })).join(' ');
        });
        return ['sm', 'md', 'lg', 'xl'].reduce((/**
         * @param {?} pre
         * @param {?} current
         * @return {?}
         */
        function (pre, current) {
            return typeof _this[current] === 'object' ? pre + " " + makeClass(current) :
                typeof _this[current] === 'number' ? pre + " col-" + current + "-" + _this[current] : pre + " ";
        }), '');
    };
    /* 获取offset */
    /* 获取offset */
    /**
     * @return {?}
     */
    FarrisColDirective.prototype.getOffsetClass = /* 获取offset */
    /**
     * @return {?}
     */
    function () {
        return this.offset ? " offset-" + this.offset : " ";
    };
    // get order class
    // get order class
    /**
     * @return {?}
     */
    FarrisColDirective.prototype.getOrderClass = 
    // get order class
    /**
     * @return {?}
     */
    function () {
        return this.order ? " order-" + this.order : " ";
    };
    FarrisColDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-col]'
                },] }
    ];
    /** @nocollapse */
    FarrisColDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    FarrisColDirective.propDecorators = {
        col: [{ type: Input }],
        colSm: [{ type: Input, args: ['col-sm',] }],
        colMd: [{ type: Input, args: ['col-md',] }],
        colLg: [{ type: Input, args: ['col-lg',] }],
        colXl: [{ type: Input, args: ['col-xl',] }],
        span: [{ type: Input }],
        offset: [{ type: Input }],
        order: [{ type: Input }],
        xs: [{ type: Input }],
        sm: [{ type: Input }],
        md: [{ type: Input }],
        lg: [{ type: Input }],
        xl: [{ type: Input }],
        ColClass: [{ type: HostBinding, args: ['class',] }],
        isCol: [{ type: HostBinding, args: ['class.col',] }],
        isColSm: [{ type: HostBinding, args: ['class.col-sm',] }],
        isColMd: [{ type: HostBinding, args: ['class.col-md',] }],
        isColLg: [{ type: HostBinding, args: ['class.col-lg',] }],
        isColXl: [{ type: HostBinding, args: ['class.col-xl',] }]
    };
    return FarrisColDirective;
}());
export { FarrisColDirective };
if (false) {
    /** @type {?} */
    FarrisColDirective.prototype.col;
    /** @type {?} */
    FarrisColDirective.prototype.colSm;
    /** @type {?} */
    FarrisColDirective.prototype.colMd;
    /** @type {?} */
    FarrisColDirective.prototype.colLg;
    /** @type {?} */
    FarrisColDirective.prototype.colXl;
    /** @type {?} */
    FarrisColDirective.prototype.span;
    /** @type {?} */
    FarrisColDirective.prototype.offset;
    /** @type {?} */
    FarrisColDirective.prototype.order;
    /** @type {?} */
    FarrisColDirective.prototype.xs;
    /** @type {?} */
    FarrisColDirective.prototype.sm;
    /** @type {?} */
    FarrisColDirective.prototype.md;
    /** @type {?} */
    FarrisColDirective.prototype.lg;
    /** @type {?} */
    FarrisColDirective.prototype.xl;
    /** @type {?} */
    FarrisColDirective.prototype.nativeClass;
    /**
     * @type {?}
     * @private
     */
    FarrisColDirective.prototype.el;
}
