// @farris/mobile-vue-adapter/esm2015/lib/lookup-mapping.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/lookup-mapping.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 帮助映射服务
 */
export class LookupMappingService {
    /**
     * @param {?} viewModelContext
     */
    constructor(viewModelContext) {
        this.viewModelContext = viewModelContext;
    }
    /**
     * 映射字段
     * \@summary
     * mapFields格式形如：
     * {
     *    id: "assoField.assoField",
     *    code: "assoField.assoField_Code",
     *    name: "assoField.assoField_Name"
     * }
     * @param {?} helpData
     * @param {?} mappingFields
     * @return {?}
     */
    mapFields(helpData, mappingFields) {
        console.log(helpData);
        if (!mappingFields) {
            return;
        }
        /** @type {?} */
        let helpFields = Object.keys(mappingFields);
        /** @type {?} */
        const idIndex = helpFields.findIndex((/**
         * @param {?} item
         * @return {?}
         */
        (item) => item === 'id'));
        if (helpFields.includes('id') && idIndex !== 0) {
            helpFields.splice(idIndex, 1);
            helpFields = ['id', ...helpFields];
        }
        /** @type {?} */
        const that = this;
        helpFields.forEach((/**
         * @param {?} f
         * @return {?}
         */
        (f) => {
            // 1、获取字段值
            // 如果helpData有选中值，则获取帮助数据源里对应你字段的值；
            // 如果helpData没有值（清空场景），则返回一个空字符串
            /** @type {?} */
            let val = '';
            if (helpData) {
                if (helpData instanceof Array) {
                    val = helpData
                        .map((/**
                     * @param {?} h
                     * @return {?}
                     */
                    (h) => {
                        return this.getValue(f, h);
                    }))
                        .join(',');
                }
                else {
                    val = that.getValue(f, helpData);
                }
            }
            // 2、设置字段值
            // 如果helpData不存在（清空场景），获取BindingData里对应字段的值，如果是数值，则设置0，其他设置上一步中的空字符串；
            // 如果helpData存在：直接设置上一步中获取的值。
            /** @type {?} */
            const pathArr = that.getBindingPathArray();
            mappingFields[f].split(',').forEach((/**
             * @param {?} ff
             * @return {?}
             */
            (ff) => {
                if (!helpData) {
                }
                else {
                    that.viewModelContext.bindingData.setValue(pathArr.concat(ff.split('.')), val, true, true);
                }
            }));
        }));
    }
    /**
     * 获取字段值
     * @param {?} fieldPath
     * @return {?}
     */
    getFieldValue(fieldPath) {
        /** @type {?} */
        const bindingPathArray = this.getBindingPathArray();
        /** @type {?} */
        const fieldPathArray = bindingPathArray.concat(fieldPath.split('.'));
        return this.viewModelContext.bindingData.getValue(fieldPathArray);
    }
    /**
     * @private
     * @param {?} f
     * @param {?} data
     * @return {?}
     */
    getValue(f, data) {
        /** @type {?} */
        let val = '';
        if (f.indexOf('.') === -1) {
            val = data[f];
        }
        else {
            val = f.split('.').reduce((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            (a, b) => {
                return a[b];
            }), data);
        }
        return val;
    }
    /**
     * @private
     * @return {?}
     */
    getBindingPathArray() {
        /** @type {?} */
        const path = this.viewModelContext.bindingData.bindingPath;
        if (path) {
            return path.split('/').filter((/**
             * @param {?} n
             * @return {?}
             */
            (n) => n !== ''));
        }
        return [];
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    LookupMappingService.prototype.viewModelContext;
}
