// @farris/ui-editor/esm5/lib/utils/ScriptLoader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { uuid } from './Utils';
/**
 * @record
 */
export function IStateObj() { }
if (false) {
    /** @type {?} */
    IStateObj.prototype.listeners;
    /** @type {?} */
    IStateObj.prototype.scriptId;
    /** @type {?} */
    IStateObj.prototype.scriptLoaded;
}
/** @type {?} */
var createState = (/**
 * @return {?}
 */
function () {
    return {
        listeners: [],
        scriptId: uuid('tiny-script'),
        scriptLoaded: false
    };
});
var ɵ0 = createState;
// WARNING: interface has both a type and a value, skipping emit
/** @type {?} */
var CreateScriptLoader = (/**
 * @return {?}
 */
function () {
    /** @type {?} */
    var state = createState();
    /** @type {?} */
    var injectScriptTag = (/**
     * @param {?} scriptId
     * @param {?} doc
     * @param {?} url
     * @param {?} callback
     * @return {?}
     */
    function (scriptId, doc, url, callback) {
        /** @type {?} */
        var scriptTag = doc.createElement('script');
        scriptTag.referrerPolicy = 'origin';
        scriptTag.type = 'application/javascript';
        scriptTag.id = scriptId;
        scriptTag.src = url;
        /** @type {?} */
        var handler = (/**
         * @return {?}
         */
        function () {
            scriptTag.removeEventListener('load', handler);
            callback();
        });
        scriptTag.addEventListener('load', handler);
        if (doc.head) {
            doc.head.appendChild(scriptTag);
        }
    });
    /** @type {?} */
    var load = (/**
     * @param {?} doc
     * @param {?} url
     * @param {?} callback
     * @return {?}
     */
    function (doc, url, callback) {
        if (state.scriptLoaded) {
            callback();
        }
        else {
            state.listeners.push(callback);
            if (!doc.getElementById(state.scriptId)) {
                injectScriptTag(state.scriptId, doc, url, (/**
                 * @return {?}
                 */
                function () {
                    state.listeners.forEach((/**
                     * @param {?} fn
                     * @return {?}
                     */
                    function (fn) { return fn(); }));
                    state.scriptLoaded = true;
                }));
            }
        }
    });
    // Only to be used by tests.
    /** @type {?} */
    var reinitialize = (/**
     * @return {?}
     */
    function () {
        state = createState();
    });
    return {
        load: load,
        reinitialize: reinitialize
    };
});
var ɵ1 = CreateScriptLoader;
/** @type {?} */
var ScriptLoader = CreateScriptLoader();
export { ScriptLoader };
export { ɵ0, ɵ1 };
