// @farris/ui-editor/esm5/lib/utils/Utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { validEvents } from '../Events';
/** @type {?} */
var bindHandlers = (/**
 * @param {?} ctx
 * @param {?} editor
 * @return {?}
 */
function (ctx, editor) {
    validEvents.forEach((/**
     * @param {?} eventName
     * @return {?}
     */
    function (eventName) {
        /** @type {?} */
        var eventEmitter = ctx[eventName];
        editor.on(eventName.substring(2), (/**
         * @param {?} event
         * @return {?}
         */
        function (event) { return ctx.ngZone.run((/**
         * @return {?}
         */
        function () { return eventEmitter.emit({ event: event, editor: editor }); })); }));
    }));
});
var ɵ0 = bindHandlers;
/** @type {?} */
var unique = 0;
/** @type {?} */
var uuid = (/**
 * @param {?} prefix
 * @return {?}
 */
function (prefix) {
    /** @type {?} */
    var date = new Date();
    /** @type {?} */
    var time = date.getTime();
    /** @type {?} */
    var random = Math.floor(Math.random() * 1000000000);
    unique++;
    return prefix + '_' + random + unique + String(time);
});
var ɵ1 = uuid;
/** @type {?} */
var isTextarea = (/**
 * @param {?=} element
 * @return {?}
 */
function (element) {
    return typeof element !== 'undefined' && element.tagName.toLowerCase() === 'textarea';
});
var ɵ2 = isTextarea;
/** @type {?} */
var normalizePluginArray = (/**
 * @param {?=} plugins
 * @return {?}
 */
function (plugins) {
    if (typeof plugins === 'undefined' || plugins === '') {
        return [];
    }
    return Array.isArray(plugins) ? plugins : plugins.split(' ');
});
var ɵ3 = normalizePluginArray;
/** @type {?} */
var mergePlugins = (/**
 * @param {?} initPlugins
 * @param {?=} inputPlugins
 * @return {?}
 */
function (initPlugins, inputPlugins) {
    return normalizePluginArray(initPlugins).concat(normalizePluginArray(inputPlugins));
});
var ɵ4 = mergePlugins;
// tslint:disable-next-line:no-empty
/** @type {?} */
var noop = (/**
 * @return {?}
 */
function () { });
var ɵ5 = noop;
/** @type {?} */
var isNullOrUndefined = (/**
 * @param {?} value
 * @return {?}
 */
function (value) { return value === null || value === undefined; });
var ɵ6 = isNullOrUndefined;
export { bindHandlers, uuid, isTextarea, normalizePluginArray, mergePlugins, noop, isNullOrUndefined };
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6 };
