// @gsp-dip/data-imp-exp/esm2015/lib/service/DataIEService/custom-data-ie.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable, Output } from '@angular/core';
import { DataIEUtils, getQueryString, isNullOrEmpty, isNullOrUndefined } from '../Utils/DataIEUtils';
import { NotifyService } from '@farris/ui-notify';
import { FileOperateService } from './file-operate.service';
import { DataImportService } from './data-import.service';
import { LoadingService } from '@farris/ui-loading';
import { HttpClient } from '@angular/common/http';
import * as i0 from "@angular/core";
import * as i1 from "@farris/ui-notify";
import * as i2 from "./file-operate.service";
import * as i3 from "@farris/ui-loading";
import * as i4 from "./data-import.service";
import * as i5 from "@angular/common/http";
export class CustomDataIEService {
    /**
     * @param {?} notify
     * @param {?} operateService
     * @param {?} loadingService
     * @param {?} importService
     * @param {?} http
     */
    constructor(notify, operateService, loadingService, importService, http) {
        this.notify = notify;
        this.operateService = operateService;
        this.loadingService = loadingService;
        this.importService = importService;
        this.http = http;
        this.origin = window.location.origin;
        this.customExportUrl = this.origin + '/api/runtime/dip/v1.0/customexport';
        this.customImportUrl = this.origin + '/api/runtime/dip/v1.0/customimport';
        this.export = new EventEmitter();
        this.complete = new EventEmitter();
        DataIEUtils.getTranslateValue(http).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            this.i18nResource = res;
        }));
    }
    /**
     * @param {?} model
     * @param {?} components
     * @param {?} frameContext
     * @param {?} suffix
     * @param {?} isImportForm
     * @param {?} i18nId
     * @param {?} dimension
     * @return {?}
     */
    SaveCustomImportRule(model, components, frameContext, suffix, isImportForm, i18nId, dimension) {
        /** @type {?} */
        const befRestService = DataIEUtils.GetBefRestService(frameContext.repository);
        /** @type {?} */
        const tempUrl = DataIEUtils.GetPreUrl(befRestService.baseUri);
        /** @type {?} */
        const su = DataIEUtils.getSu(tempUrl);
        /** @type {?} */
        let importParam = null;
        if (isNullOrUndefined(dimension)) {
            importParam = {
                body: {
                    dataExchangeModel: JSON.stringify(model),
                    components: components,
                    suffix: suffix,
                    keys: ['DataImport', model['MainObject']['Source']['SourceMetadataID']],
                    importForm: isImportForm,
                    su: su,
                    i18nId: i18nId
                }
            };
        }
        else {
            importParam = {
                body: {
                    dataExchangeModel: JSON.stringify(model),
                    components: components,
                    suffix: suffix,
                    keys: ['DataImport', model['MainObject']['Source']['SourceMetadataID'], dimension],
                    importForm: isImportForm,
                    su: su,
                    i18nId: i18nId
                }
            };
        }
        /** @type {?} */
        let subject;
        if (!isNullOrUndefined(this.http)) {
            subject = this.http.post(`${this.customImportUrl}/AddImportRule`, importParam.body, { params: { su: su } });
        }
        else {
            subject = befRestService.request(`${this.customImportUrl}/AddImportRule`, 'post', { su: su }, importParam);
        }
        return subject;
    }
    /**
     * @param {?} model
     * @param {?} components
     * @param {?} frameContext
     * @param {?} suffix
     * @param {?} i18nId
     * @param {?} dimension
     * @return {?}
     */
    SaveCustomExportRule(model, components, frameContext, suffix, i18nId, dimension) {
        /** @type {?} */
        const befRestService = DataIEUtils.GetBefRestService(frameContext.repository);
        /** @type {?} */
        const tempUrl = DataIEUtils.GetPreUrl(befRestService.baseUri);
        /** @type {?} */
        const su = DataIEUtils.getSu(tempUrl);
        /** @type {?} */
        let exportParam = null;
        if (isNullOrUndefined(dimension)) {
            exportParam = {
                body: {
                    dataExchangeModel: JSON.stringify(model),
                    components: components,
                    suffix: suffix,
                    keys: ['DataExport', model['MainObject']['Source']['SourceMetadataID']],
                    su: su,
                    i18nId: i18nId
                }
            };
        }
        else {
            exportParam = {
                body: {
                    dataExchangeModel: JSON.stringify(model),
                    components: components,
                    suffix: suffix,
                    keys: ['DataExport', model['MainObject']['Source']['SourceMetadataID'], dimension],
                    su: su,
                    i18nId: i18nId
                }
            };
        }
        /** @type {?} */
        let subject;
        if (!isNullOrUndefined(this.http)) {
            subject = this.http.post(`${this.customExportUrl}/AddExportRule`, exportParam.body, { params: { su: su } });
        }
        else {
            subject = befRestService.request(`${this.customExportUrl}/AddExportRule`, 'post', { su: su }, exportParam);
        }
        return subject;
    }
    // 下载模板
    /**
     * @param {?} model
     * @param {?} frameContext
     * @param {?} importOption
     * @return {?}
     */
    DowmloadTemplate(model, frameContext, importOption) {
        /** @type {?} */
        const keys = ['DataImport', model['MainObject']['Source']['SourceMetadataID'], importOption.dimension];
        /** @type {?} */
        const befRestService = DataIEUtils.GetBefRestService(frameContext.repository);
        /** @type {?} */
        const tempUrl = DataIEUtils.GetPreUrl(befRestService.baseUri);
        /** @type {?} */
        const su = DataIEUtils.getSu(tempUrl);
        /** @type {?} */
        const templateParam = {
            body: {
                keys: keys,
                su: su
            }
        };
        /** @type {?} */
        const subject = befRestService.request(`${this.customImportUrl}/TemplateDownload`, 'post', { su: su }, templateParam);
        subject.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            if (data['success'] === true) {
                /** @type {?} */
                let filename = data['fileName'];
                if (importOption !== null && importOption['templatefilename'] !== null && importOption['templatefilename'] !== undefined) {
                    /** @type {?} */
                    const suffixIndex = filename.lastIndexOf('.');
                    /** @type {?} */
                    const suffix = filename.substring(suffixIndex);
                    filename = importOption['templatefilename'] + suffix;
                }
                /** @type {?} */
                const preUrl = DataIEUtils.GetPreUrl(befRestService.baseUri);
                this.operateService.DownloadFile(data['relativePath'], filename, preUrl, importOption);
                /** @type {?} */
                const subscription = this.operateService.downloadComplete.subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                result => {
                    if (result === '') {
                        this.complete.emit('');
                        subscription.unsubscribe();
                    }
                    else {
                        this.complete.emit(result);
                        subscription.unsubscribe();
                    }
                }));
            }
            else {
                this.complete.emit(data['message']);
            }
        }));
    }
    /**
     * @param {?} keys
     * @param {?} repository
     * @param {?} i18nId
     * @param {?} model
     * @return {?}
     */
    GetImportRuleModel(keys, repository, i18nId, model) {
        /** @type {?} */
        const befRestService = DataIEUtils.GetBefRestService(repository);
        /** @type {?} */
        const tempUrl = DataIEUtils.GetPreUrl(befRestService.baseUri);
        /** @type {?} */
        const su = DataIEUtils.getSu(tempUrl);
        /** @type {?} */
        const params = {
            body: {
                keys: keys,
                su: su,
                i18nId: i18nId,
                needI18nModel: JSON.stringify(model)
            }
        };
        /** @type {?} */
        let subject;
        if (!isNullOrUndefined(this.http)) {
            subject = this.http.post(`${this.customImportUrl}/GetImportModel`, params.body, { params: { su: su } });
        }
        else {
            subject = befRestService.request(`${this.customImportUrl}/GetImportModel`, 'post', { su: su }, params);
        }
        return subject;
    }
    /**
     * @param {?} keys
     * @param {?} repository
     * @param {?} i18nId
     * @param {?} model
     * @return {?}
     */
    GetExportRuleModel(keys, repository, i18nId, model) {
        /** @type {?} */
        const befRestService = DataIEUtils.GetBefRestService(repository);
        /** @type {?} */
        const tempUrl = DataIEUtils.GetPreUrl(befRestService.baseUri);
        /** @type {?} */
        const su = DataIEUtils.getSu(tempUrl);
        /** @type {?} */
        const params = {
            body: {
                keys: keys,
                su: su,
                i18nId: i18nId,
                needI18nModel: JSON.stringify(model)
            }
        };
        /** @type {?} */
        let subject;
        if (!isNullOrUndefined(this.http)) {
            subject = this.http.post(`${this.customExportUrl}/GetExportModel`, params.body, { params: { su: su } });
        }
        else {
            subject = befRestService.request(`${this.customExportUrl}/GetExportModel`, 'post', { su: su }, params);
        }
        return subject;
    }
    /**
     * @param {?} commonParams
     * @return {?}
     */
    ImportByRule(commonParams) {
        /** @type {?} */
        const repository = commonParams.repository;
        /** @type {?} */
        const option = commonParams.option;
        /** @type {?} */
        const keys = commonParams.keys;
        /** @type {?} */
        const fileName = commonParams.fileName;
        /** @type {?} */
        const befRestService = DataIEUtils.GetBefRestService(repository);
        /** @type {?} */
        let globalVar = '';
        if (option !== null && option['GlobalParam'] !== null && option['GlobalParam'] !== undefined) {
            globalVar = JSON.stringify(option['GlobalParam']);
        }
        /** @type {?} */
        const funcId = getQueryString('funcId');
        /** @type {?} */
        const importParam = {
            body: {
                dataImportContext: {
                    Keys: keys,
                    FileName: fileName,
                    GlobalParam: globalVar,
                    funcId: funcId
                }
            }
        };
        commonParams.befRestService = befRestService;
        commonParams.importParam = importParam;
        commonParams.mainObjId = '';
        this.importService.Import(commonParams);
    }
    /**
     * @param {?} commonParams
     * @return {?}
     */
    ImportSubTableByRule(commonParams) {
        /** @type {?} */
        const repository = commonParams.repository;
        /** @type {?} */
        const option = commonParams.option;
        /** @type {?} */
        const keys = commonParams.keys;
        /** @type {?} */
        const fileName = commonParams.fileName;
        /** @type {?} */
        const frameContext = commonParams.frameContext;
        /** @type {?} */
        const befRestService = DataIEUtils.GetBefRestService(repository);
        /** @type {?} */
        let globalVar = '';
        if (option !== null && option['GlobalParam'] !== null && option['GlobalParam'] !== undefined) {
            globalVar = JSON.stringify(option['GlobalParam']);
        }
        if (frameContext.bindingData.list.currentId === '' || frameContext.bindingData.list.currentId === null) {
            this.notify.info(this.i18nResource['master_table_noid']);
            return false;
        }
        else {
            /** @type {?} */
            const funcId = getQueryString('funcId');
            /** @type {?} */
            const importParam = {
                body: {
                    dataImportContext: {
                        Keys: keys,
                        FileName: fileName,
                        MainObjId: frameContext.bindingData.list.currentId,
                        GlobalParam: globalVar,
                        funcId: funcId
                    },
                    requestInfo: befRestService.buildRequestInfo()
                }
            };
            commonParams.befRestService = befRestService;
            commonParams.importParam = importParam;
            commonParams.mainObjId = frameContext.bindingData.list.currentId;
            this.importService.Import(commonParams);
        }
    }
    /**
     * @param {?} keys
     * @param {?} filter
     * @param {?} option
     * @param {?} frameContext
     * @return {?}
     */
    ExportByRule(keys, filter, option, frameContext) {
        /** @type {?} */
        const befRestService = DataIEUtils.GetBefRestService(frameContext.repository);
        /** @type {?} */
        const funcId = getQueryString('funcId');
        /** @type {?} */
        const newFilters = [];
        /** @type {?} */
        const selectFilterGridComponent = option['selectFilterGridComponent'];
        if (!isNullOrUndefined(selectFilterGridComponent) && !isNullOrEmpty(selectFilterGridComponent)) {
            /** @type {?} */
            const ids = frameContext.appContext.frameContextManager.getFrameContextById(selectFilterGridComponent).uiState['ids'];
            if (!isNullOrUndefined(ids) && ids.length > 0) {
                /** @type {?} */
                let idsStr = '';
                for (let i = 0; i < ids.length; i++) {
                    /** @type {?} */
                    const id = ids[i];
                    if (i === ids.length - 1) {
                        idsStr += id;
                    }
                    else {
                        idsStr += id + '\r\n';
                    }
                }
                /** @type {?} */
                const selectIdFilterCondition = {
                    FilterField: "ID",
                    Compare: 14,
                    Value: idsStr,
                    Relation: 1,
                    Expresstype: 0
                };
                newFilters.push(selectIdFilterCondition);
            }
            else {
                this.notify.info(this.i18nResource['select-data-export']);
                return;
            }
        }
        if (option['enableScreeningScheme'] === true) {
            /** @type {?} */
            const filterConditionList = frameContext.uiState['filterConditionList'];
            if (filterConditionList !== undefined && filterConditionList !== null && filterConditionList !== '' && filterConditionList !== '[]') {
                /** @type {?} */
                let filterConditionListJson;
                if (typeof (filterConditionList) === 'string') {
                    filterConditionListJson = JSON.parse(filterConditionList);
                }
                else {
                    filterConditionListJson = filterConditionList;
                }
                for (let index = 0; index < filterConditionListJson.length; index++) {
                    /** @type {?} */
                    const element = filterConditionListJson[index];
                    /** @type {?} */
                    const value = decodeURIComponent(element.Value);
                    element.Value = value;
                    newFilters.push(element);
                }
            }
        }
        if (newFilters.length > 0) {
            if (!isNullOrUndefined(filter) && filter.FilterConditions !== null) {
                filter.FilterConditions = newFilters.concat(filter.FilterConditions);
            }
            else {
                if (isNullOrUndefined(filter)) {
                    filter = {};
                }
                filter['FilterConditions'] = newFilters;
            }
        }
        if (!isNullOrUndefined(filter) && !isNullOrUndefined(filter.FilterConditions)) {
            //确保最后一个关系为空
            filter.FilterConditions[filter.FilterConditions.length - 1].Relation = 0;
        }
        /** @type {?} */
        const exportParam = {
            body: {
                exportContext: {
                    Filter: filter,
                    Keys: keys,
                    GlobalParam: '',
                    funcId: funcId
                }
            }
        };
        this.loadingService.show();
        /** @type {?} */
        const subject = befRestService.request(`${befRestService.baseUri}/service/dataexportvmaction`, 'put', null, exportParam);
        subject.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            if (data === null) {
                // this.export.emit(false);
                this.notify.info('数据导出失败！\n 失败原因：服务端响应异常，请检查');
            }
            /** @type {?} */
            let exportResult = (/** @type {?} */ (data.returnValue));
            if (isNullOrUndefined(exportResult)) {
                exportResult = (/** @type {?} */ (data));
            }
            /** @type {?} */
            let fileName = exportResult.fileName;
            if (option !== null && option['customfilename'] !== null && option['customfilename'] !== undefined) {
                /** @type {?} */
                const suffixIndex = fileName.lastIndexOf('.');
                /** @type {?} */
                const suffix = fileName.substring(suffixIndex);
                fileName = option['customfilename'] + suffix;
            }
            if (exportResult.succeed === true) {
                /** @type {?} */
                const preUrl = DataIEUtils.GetPreUrl(befRestService.baseUri);
                this.operateService.DownloadFile(exportResult.relativePath, fileName, preUrl, option);
                /** @type {?} */
                const subscriber = this.operateService.downloadComplete.subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                result => {
                    this.loadingService.clearAll();
                    if (result === '') {
                        this.export.emit('');
                        subscriber.unsubscribe();
                    }
                    else {
                        this.export.emit(result);
                        subscriber.unsubscribe();
                    }
                }));
            }
            else {
                this.loadingService.clearAll();
                this.errorlevel = exportResult.errorLevel;
                this.export.emit(exportResult.message);
                // this.notify.error('数据导出失败！\n 失败原因：' + exportResult.message);
            }
        }));
    }
}
CustomDataIEService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
CustomDataIEService.ctorParameters = () => [
    { type: NotifyService },
    { type: FileOperateService },
    { type: LoadingService },
    { type: DataImportService },
    { type: HttpClient }
];
CustomDataIEService.propDecorators = {
    export: [{ type: Output }]
};
/** @nocollapse */ CustomDataIEService.ngInjectableDef = i0.defineInjectable({ factory: function CustomDataIEService_Factory() { return new CustomDataIEService(i0.inject(i1.NotifyService), i0.inject(i2.FileOperateService), i0.inject(i3.LoadingService), i0.inject(i4.DataImportService), i0.inject(i5.HttpClient)); }, token: CustomDataIEService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    CustomDataIEService.prototype.origin;
    /**
     * @type {?}
     * @private
     */
    CustomDataIEService.prototype.customExportUrl;
    /**
     * @type {?}
     * @private
     */
    CustomDataIEService.prototype.customImportUrl;
    /** @type {?} */
    CustomDataIEService.prototype.complete;
    /** @type {?} */
    CustomDataIEService.prototype.export;
    /** @type {?} */
    CustomDataIEService.prototype.errorlevel;
    /** @type {?} */
    CustomDataIEService.prototype.i18nResource;
    /**
     * @type {?}
     * @private
     */
    CustomDataIEService.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    CustomDataIEService.prototype.operateService;
    /**
     * @type {?}
     * @private
     */
    CustomDataIEService.prototype.loadingService;
    /**
     * @type {?}
     * @private
     */
    CustomDataIEService.prototype.importService;
    /**
     * @type {?}
     * @private
     */
    CustomDataIEService.prototype.http;
}
