// @gsp-dip/data-imp-exp/esm2015/lib/service/DataIEService/data-import.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ComponentFactoryResolver, EventEmitter, Injectable, Injector } from '@angular/core';
import { BsModalService } from '@farris/ui-modal';
import { ValidateFormComponent } from '../../validate-form/validate-form.component';
import { FileOperateService } from './file-operate.service';
import { NotifyService } from '@farris/ui-notify';
import { DataIEUtils, getQueryString, isNullOrEmpty, isNullOrUndefined } from '../Utils/DataIEUtils';
import { HttpClient } from '@angular/common/http';
import { MethodType } from "../../model/DataIEType/MethodType";
export class DataImportService {
    /**
     * @param {?} modalService
     * @param {?} resolver
     * @param {?} operateService
     * @param {?} notify
     * @param {?} http
     * @param {?} injector
     */
    constructor(modalService, resolver, operateService, notify, http, injector) {
        this.modalService = modalService;
        this.resolver = resolver;
        this.operateService = operateService;
        this.notify = notify;
        this.http = http;
        this.injector = injector;
        this.origin = window.location.origin;
        this.url = this.origin + '/api/runtime/rtcommon/v1.0/import';
        this.DataIeProgressParams = {
            "ImportProgressValue": 13
        };
        this.complete = new EventEmitter();
        DataIEUtils.getTranslateValue(http).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            this.master_table_noid = res['master_table_noid'];
            this.bind_form_error = res['bind_form_error'];
            this.validate_result_show = res['validate_result_show'];
        }));
    }
    /**
     * @param {?} commonParams
     * @return {?}
     */
    ImportSubTable(commonParams) {
        /** @type {?} */
        const ruleId = commonParams.ruleId;
        /** @type {?} */
        const option = commonParams.option;
        /** @type {?} */
        const frameContext = commonParams.frameContext;
        /** @type {?} */
        const methodType = commonParams.methodType;
        /** @type {?} */
        const suffix = commonParams.suffix;
        /** @type {?} */
        const fileName = commonParams.fileName;
        /** @type {?} */
        const befRestService = this.getBefRestService(frameContext.repository);
        commonParams.befRestService = befRestService;
        commonParams.repository = frameContext.repository;
        /** @type {?} */
        let globalVar = '';
        if (option['GlobalParam'] !== null && option['GlobalParam'] !== undefined) {
            if (typeof (option['GlobalParam']) === 'object') {
                globalVar = JSON.stringify(option['GlobalParam']);
            }
            else {
                globalVar = option['GlobalParam'];
            }
        }
        /** @type {?} */
        let model = '';
        if (option.Models !== undefined && option.Models !== null && option.Models.length > 0) {
            model = JSON.stringify(option.Models[0]);
        }
        /** @type {?} */
        const mainObjId = frameContext.bindingData.list.currentId;
        if (isNullOrEmpty(mainObjId)) {
            this.notify.info(this.master_table_noid);
            return false;
            // this.errorlevel = 'Info';
            // this.complete.emit('未获取到主表ID值，请确认主表是否有数据');
        }
        else {
            /** @type {?} */
            const funcId = getQueryString('funcId');
            /** @type {?} */
            const importParam = {
                body: {
                    dataImportContext: {
                        RuleId: ruleId,
                        FileName: fileName,
                        MainObjId: mainObjId,
                        GlobalParam: globalVar,
                        Suffix: suffix,
                        Model: model,
                        funcId: funcId,
                        securityLevelName: commonParams.securityLevelName,
                        methodType: methodType
                    },
                    requestInfo: befRestService.buildRequestInfo()
                }
            };
            commonParams.importParam = importParam;
            commonParams.mainObjId = mainObjId;
            this.Import(commonParams);
        }
    }
    /**
     * @param {?} commonParams
     * @return {?}
     */
    ImportSubTable4HandCraft(commonParams) {
        /** @type {?} */
        const ruleId = commonParams.ruleId;
        /** @type {?} */
        const option = commonParams.option;
        /** @type {?} */
        const frameContext = commonParams.frameContext;
        /** @type {?} */
        const suffix = commonParams.suffix;
        /** @type {?} */
        const fileName = commonParams.fileName;
        /** @type {?} */
        let globalVar = '';
        if (option['GlobalParam'] !== null && option['GlobalParam'] !== undefined) {
            if (typeof (option['GlobalParam']) === 'object') {
                globalVar = JSON.stringify(option['GlobalParam']);
            }
            else {
                globalVar = option['GlobalParam'];
            }
        }
        /** @type {?} */
        const repository = frameContext === null ? null : frameContext.repository;
        /** @type {?} */
        let mainObjId = '';
        //手工表单仓库为null
        if (repository === null) {
            if (option['mainId'] === undefined || option['mainId'] === null || option['mainId'] === '') {
                this.notify.info(this.master_table_noid);
                return false;
            }
            else {
                mainObjId = option['mainId'];
            }
        }
        else { //通用导入
            if (frameContext.bindingData.list.currentId === '' || frameContext.bindingData.list.currentId === null) {
                this.notify.info(this.master_table_noid);
                return false;
            }
            else {
                mainObjId = frameContext.bindingData.list.currentId;
            }
        }
        /** @type {?} */
        const funcId = getQueryString('funcId');
        /** @type {?} */
        const importParam = {
            ruleId: ruleId,
            fileName: fileName,
            mainObjId: mainObjId,
            globalParam: globalVar,
            suffix: suffix,
            funcId: funcId,
            securityLevelName: commonParams.securityLevelName
        };
        commonParams.importParam = importParam;
        commonParams.mainObjId = mainObjId;
        // this.Import4HandCraft(commonParams);
    }
    /**
     * @param {?} commonParams
     * @return {?}
     */
    ImportByRuleId(commonParams) {
        /** @type {?} */
        const ruleId = commonParams.ruleId;
        /** @type {?} */
        const option = commonParams.option;
        /** @type {?} */
        const repository = commonParams.repository;
        /** @type {?} */
        const suffix = commonParams.suffix;
        /** @type {?} */
        const fileName = commonParams.fileName;
        /** @type {?} */
        const methodType = commonParams.methodType;
        /** @type {?} */
        const befRestService = this.getBefRestService(repository);
        commonParams.befRestService = befRestService;
        /** @type {?} */
        let globalVar = '';
        if (option['GlobalParam'] !== null && option['GlobalParam'] !== undefined) {
            if (typeof (option['GlobalParam']) === 'object') {
                globalVar = JSON.stringify(option['GlobalParam']);
            }
            else {
                globalVar = option['GlobalParam'];
            }
        }
        /** @type {?} */
        let model = '';
        if (option.Models !== undefined && option.Models !== null && option.Models.length > 0) {
            model = JSON.stringify(option.Models[0]);
        }
        console.log('baseUri: ' + befRestService.baseUri);
        /** @type {?} */
        const funcId = getQueryString('funcId');
        if (methodType === MethodType.CardFormDataImport) {
            commonParams.mainObjId = option.mainObjId;
        }
        else {
            commonParams.mainObjId = '';
        }
        /** @type {?} */
        const importParam = {
            body: {
                dataImportContext: {
                    RuleId: ruleId,
                    FileName: fileName,
                    GlobalParam: globalVar,
                    Suffix: suffix,
                    Model: model,
                    funcId: funcId,
                    securityLevelName: commonParams.securityLevelName,
                    mainObjId: commonParams.mainObjId,
                    methodType: methodType
                }
            }
        };
        commonParams.importParam = importParam;
        this.Import(commonParams);
    }
    /**
     * @param {?} commonParams
     * @return {?}
     */
    Import(commonParams) {
        /** @type {?} */
        const repository = commonParams.repository;
        /** @type {?} */
        const importParam = commonParams.importParam;
        /** @type {?} */
        const befRestService = commonParams.befRestService;
        /** @type {?} */
        const methodType = commonParams.methodType;
        /** @type {?} */
        const subscriber = befRestService.request(`${befRestService.baseUri}/service/dataimportvmaction`, 'put', null, importParam);
        subscriber.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            /** @type {?} */
            const su = DataIEUtils.getSu(befRestService.baseUri);
            /** @type {?} */
            let importResult = (/** @type {?} */ (data.returnValue));
            if (isNullOrUndefined(importResult)) {
                importResult = (/** @type {?} */ (data));
            }
            commonParams.su = su;
            commonParams.importResult = importResult;
            if (importResult.succeed === true) {
                try {
                    if (importResult.formData && importResult.formData.length > 0) {
                        importResult.isform = true;
                        if (methodType === MethodType.ImportSubTableDetail) {
                            //子表导入表单
                            /** @type {?} */
                            const newData = importResult.formData[0];
                            /** @type {?} */
                            const curEntity = repository.entityCollection.getEntityById(commonParams.mainObjId);
                            curEntity.load(newData);
                        }
                        else if (methodType === MethodType.CardFormDataImport) {
                            //卡片导入不走数据加载，导入命令完成后有更新命令
                        }
                        else {
                            //主从表导入表单
                            /** @type {?} */
                            const entities = [];
                            importResult.formData.forEach((/**
                             * @param {?} rdata
                             * @return {?}
                             */
                            (rdata) => {
                                /** @type {?} */
                                const entity = repository.buildEntity(rdata);
                                entities.push(entity);
                            }));
                            repository.entityCollection.addEntities(entities);
                        }
                    }
                    if (importResult.validatedResults && importResult.validatedResults.length > 0) {
                        this.complete.emit('部分数据导入完成');
                    }
                    else {
                        if (importResult.isform == true) {
                            //此处为传递，无需国际化
                            this.complete.emit('导入表单成功');
                        }
                        else {
                            this.complete.emit('');
                        }
                    }
                    // this.complete.emit(true);
                    //this.complete.emit('');
                    //导入成功如果展示校验结果，只能是校验策略是忽略，所有校验结果都是error
                    this.ShowValidatedResult(commonParams);
                    // this.notify.info('数据导入成功');
                    // this.needDestroy.emit();
                }
                catch (_a) {
                    // this.complete.emit(false);
                    this.complete.emit(importResult.message);
                    this.notify.error(this.bind_form_error);
                }
            }
            else {
                console.log(importResult.exceptionSource);
                this.errorlevel = importResult.errorLevel;
                this.ShowValidatedResult(commonParams);
                // this.complete.emit(importResult.message);
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.log(error);
            this.errorlevel = 'error';
            this.complete.emit(error.error.Message);
        }));
    }
    // 显示校验结果
    /**
     * @private
     * @param {?} commonParams
     * @return {?}
     */
    ShowValidatedResult(commonParams) {
        /** @type {?} */
        const importResult = commonParams.importResult;
        /** @type {?} */
        const loadingParam = commonParams.loadingParam;
        /** @type {?} */
        const befRestService = commonParams.befRestService;
        if (importResult.validatedResults && importResult.validatedResults.length > 0) {
            loadingParam.loadingService.clearAll();
            /** @type {?} */
            const cmpRef = this.CreateValidatedComponent();
            cmpRef.instance.validateResult = importResult.validatedResults;
            cmpRef.instance.su = commonParams.su;
            cmpRef.instance.i18nResource = commonParams.i18nResource;
            cmpRef.instance.loadingParam = loadingParam;
            /** @type {?} */
            const validateModalConfig = {
                title: this.validate_result_show,
                width: 950,
                height: 550,
                buttons: cmpRef.instance.modalFooter,
                showHeader: true,
                showMaxButton: false,
                showButtons: true
            };
            /** @type {?} */
            const dialog = this.modalService.show(cmpRef, validateModalConfig);
            //0代表按钮中止或X点击；1代表按钮继续
            /** @type {?} */
            let closed = 0;
            dialog.dialog.instance.closed.subscribe((/**
             * @return {?}
             */
            () => {
                if (closed === 0) {
                    this.complete.emit('校验中止');
                }
            }));
            cmpRef.instance.closeModal.subscribe((/**
             * @return {?}
             */
            () => {
                closed = 0;
                dialog.close();
            }));
            cmpRef.instance.okModal.subscribe((/**
             * @param {?} validatedInfoList
             * @return {?}
             */
            (validatedInfoList) => {
                closed = 1;
                dialog.close();
                loadingParam.loadingService.show(loadingParam.loadingConfig);
                commonParams.importParam.body.dataImportContext['validatedInfoList'] = validatedInfoList;
                this.Import(commonParams);
            }));
        }
        else {
            this.complete.emit(importResult.message);
        }
    }
    /**
     * @private
     * @param {?} repository
     * @return {?}
     */
    getBefRestService(repository) {
        /** @type {?} */
        const befRepository = (/** @type {?} */ (repository));
        return befRepository.restService;
    }
    // 生成校验窗口组件
    /**
     * @private
     * @return {?}
     */
    CreateValidatedComponent() {
        /** @type {?} */
        const compFactory = this.resolver.resolveComponentFactory(ValidateFormComponent);
        /** @type {?} */
        const comRef = compFactory.create(this.injector);
        return comRef;
    }
    /**
     * @param {?} str
     * @return {?}
     */
    JudgeIllegalChar(str) {
        /** @type {?} */
        const illegalChar = ['?', '、', '\\', '/', '*', '\'', '"', '“', '”', '<', '>', '|', ','];
        illegalChar.forEach((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            if (str.indexOf(item) !== -1) {
                return true;
            }
        }));
        return false;
    }
    // 下载模板
    /**
     * @param {?} ruleId
     * @param {?} repository
     * @param {?} importOption
     * @param {?} suffix
     * @return {?}
     */
    DowmloadTemplate(ruleId, repository, importOption, suffix) {
        /** @type {?} */
        const befRestService = this.getBefRestService(repository);
        /** @type {?} */
        const preUrl = DataIEUtils.GetPreUrl(befRestService.baseUri);
        /** @type {?} */
        const tempUrl = `${preUrl}/import`;
        /** @type {?} */
        let su = DataIEUtils.getSu(tempUrl);
        if (importOption['su'] !== undefined && importOption['su'] !== null) {
            su = importOption['su'];
        }
        /** @type {?} */
        let securityLevelName = '';
        if (!isNullOrUndefined(importOption.securityLevelName)) {
            securityLevelName = importOption.securityLevelName;
        }
        /** @type {?} */
        let model = '';
        if (importOption.Models !== undefined && importOption.Models !== null && importOption.Models.length > 0) {
            model = JSON.stringify(importOption.Models[0]);
        }
        /** @type {?} */
        let globalVar = '';
        if (!isNullOrUndefined(importOption['GlobalParam'])) {
            if (typeof (importOption['GlobalParam']) === 'object') {
                globalVar = JSON.stringify(importOption['GlobalParam']);
            }
            else {
                globalVar = importOption['GlobalParam'];
            }
        }
        /** @type {?} */
        const templateParam = {
            body: {
                ruleId: ruleId,
                suffix: suffix,
                su: su,
                model: model,
                globalParam: globalVar,
                securityLevelName: securityLevelName
            }
        };
        /** @type {?} */
        const fixedUrl = this.origin + '/api/runtime/dip/v1.0/rpcimport';
        /** @type {?} */
        const subscriber = befRestService.request(`${fixedUrl}/TemplateDownload`, 'post', { su: su }, templateParam);
        subscriber.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            if (data['result'] === true) {
                /** @type {?} */
                let filename = '';
                if (importOption['templatefilename'] === '' || importOption['templatefilename'] === undefined) {
                    filename = data['fileName'];
                }
                else {
                    /** @type {?} */
                    const suffixIndex = data['fileName'].lastIndexOf('.');
                    /** @type {?} */
                    const tempsuffix = data['fileName'].substring(suffixIndex);
                    /** @type {?} */
                    let customName = importOption['templatefilename'];
                    if (!this.JudgeIllegalChar(customName)) {
                        if (customName.indexOf('.') !== -1) {
                            filename = customName;
                        }
                        else {
                            filename = customName + tempsuffix;
                        }
                    }
                    else {
                        filename = data['fileName'];
                    }
                }
                this.operateService.DownloadFile(data['docRelativePath'], filename, preUrl, importOption);
                /** @type {?} */
                const subscription = this.operateService.downloadComplete.subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                result => {
                    if (result === '') {
                        this.complete.emit('');
                        subscription.unsubscribe();
                    }
                    else {
                        this.complete.emit(result);
                        subscription.unsubscribe();
                    }
                }));
            }
            else {
                // this.notify.error('下载模板失败，失败信息：' + data['message']);
                this.complete.emit(data['message']);
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.log(error);
            this.complete.emit('下载模板失败');
        }));
    }
    // 下载模板
    /**
     * @param {?} ruleId
     * @param {?} importOption
     * @param {?} suffix
     * @return {?}
     */
    DowmloadTemplate4HandCraft(ruleId, importOption, suffix) {
        /** @type {?} */
        const su = importOption.su;
        /** @type {?} */
        let globalVar = '';
        if (importOption['GlobalParam'] !== null && importOption['GlobalParam'] !== undefined) {
            if (typeof (importOption['GlobalParam']) === 'object') {
                globalVar = JSON.stringify(importOption['GlobalParam']);
            }
            else {
                globalVar = importOption['GlobalParam'];
            }
        }
        /** @type {?} */
        let securityLevelName = '';
        if (!isNullOrUndefined(importOption.securityLevelName)) {
            securityLevelName = importOption.securityLevelName;
        }
        /** @type {?} */
        const templateParam = {
            ruleId: ruleId,
            suffix: suffix,
            su: su,
            globalParam: globalVar,
            securityLevelName: securityLevelName
        };
        /** @type {?} */
        const fixedUrl = this.origin + '/api/runtime/dip/v1.0/AllSteps/TemplateDownload';
        this.http.post(fixedUrl, templateParam, {
            headers: {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            responseType: 'blob',
            observe: 'response',
            params: { su: su }
        }).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        (res) => {
            console.log(res);
            /** @type {?} */
            const data = res.body;
            if (data.type == 'application/json') {
                /** @type {?} */
                const reader = new FileReader();
                reader.onload = (/**
                 * @param {?} e
                 * @return {?}
                 */
                (e) => {
                    /** @type {?} */
                    const responseParam = JSON.parse((/** @type {?} */ (reader.result)));
                    if (!responseParam.result) {
                        this.complete.emit(responseParam.message);
                    }
                });
                reader.readAsText(data, 'utf-8');
            }
            else {
                /** @type {?} */
                let fileName = '';
                /** @type {?} */
                let contentDisposition = res.headers.get("content-disposition");
                if (contentDisposition) {
                    // 正则获取filename的值
                    /** @type {?} */
                    let filenameRegex = /filename[^;=\n]*=((['"]).*?\2|[^;\n]*)/;
                    /** @type {?} */
                    let matches = filenameRegex.exec(contentDisposition);
                    if (matches != null && matches[1]) {
                        fileName = matches[1].replace(/['"]/g, '');
                    }
                    // 通过 URLEncoder.encode(pFileName, StandardCharsets.UTF_8.name()) 加密编码的, 使用decodeURI(fileName) 解密
                    fileName = decodeURI(fileName);
                    // 通过 new String(pFileName.getBytes(), StandardCharsets.ISO_8859_1) 加密编码的, 使用decodeURI(escape(fileName)) 解密
                    // fileName = decodeURI(escape(fileName))
                }
                if (!isNullOrUndefined(importOption.templatefilename)) {
                    /** @type {?} */
                    const suffixIndex = fileName.lastIndexOf('.');
                    /** @type {?} */
                    const suffix = fileName.substring(suffixIndex);
                    fileName = importOption.templatefilename + suffix;
                }
                /** @type {?} */
                const blob = new Blob([data]);
                if (!data || data.size == 0) {
                    this.complete.emit('返回的数据为空');
                }
                else {
                    if (navigator.msSaveOrOpenBlob) {
                        navigator.msSaveOrOpenBlob(blob, fileName);
                    }
                    else {
                        /** @type {?} */
                        const link = document.createElement('a');
                        link.setAttribute('href', window.URL.createObjectURL(blob));
                        // 设置下载的文件名
                        link.setAttribute('download', fileName);
                        link.style.visibility = 'hidden';
                        document.body.appendChild(link);
                        link.click();
                        document.body.removeChild(link);
                    }
                    this.complete.emit('');
                }
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            this.complete.error(error.error.Message);
        }));
    }
}
DataImportService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DataImportService.ctorParameters = () => [
    { type: BsModalService },
    { type: ComponentFactoryResolver },
    { type: FileOperateService },
    { type: NotifyService },
    { type: HttpClient },
    { type: Injector }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataImportService.prototype.origin;
    /**
     * @type {?}
     * @private
     */
    DataImportService.prototype.url;
    /** @type {?} */
    DataImportService.prototype.DataIeProgressParams;
    /** @type {?} */
    DataImportService.prototype.complete;
    /** @type {?} */
    DataImportService.prototype.errorlevel;
    /** @type {?} */
    DataImportService.prototype.master_table_noid;
    /** @type {?} */
    DataImportService.prototype.bind_form_error;
    /** @type {?} */
    DataImportService.prototype.validate_result_show;
    /**
     * @type {?}
     * @private
     */
    DataImportService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    DataImportService.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    DataImportService.prototype.operateService;
    /**
     * @type {?}
     * @private
     */
    DataImportService.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    DataImportService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    DataImportService.prototype.injector;
}
