// @edp-aif/runtime-api/bundles/edp-aif-runtime-api.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/forms'), require('@angular/common/http'), require('@farris/ui-dialog'), require('@edp-aif/common-api'), require('@farris/ui-treetable'), require('@farris/ui-notify'), require('@farris/ui-messager'), require('@farris/command-services'), require('@angular/core'), require('@farris/ui-modal'), require('@farris/devkit'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@edp-aif/runtime-api', ['exports', '@angular/common', '@angular/forms', '@angular/common/http', '@farris/ui-dialog', '@edp-aif/common-api', '@farris/ui-treetable', '@farris/ui-notify', '@farris/ui-messager', '@farris/command-services', '@angular/core', '@farris/ui-modal', '@farris/devkit', 'rxjs', 'rxjs/operators'], factory) :
    (factory((global['edp-aif'] = global['edp-aif'] || {}, global['edp-aif']['runtime-api'] = {}),global.ng.common,global.ng.forms,global.ng.common.http,global.uiDialog,global.commonApi,global.uiTreetable,global.uiNotify,global.uiMessager,global.commandServices,global.ng.core,global.uiModal,global.devkit,global.rxjs,global.rxjs.operators));
}(this, (function (exports,common,forms,http,uiDialog,commonApi,uiTreetable,uiNotify,uiMessager,commandServices,core,uiModal,devkit,rxjs,operators) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m)
            return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length)
                    o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MappingAssociation = /** @class */ (function (_super) {
        __extends(MappingAssociation, _super);
        function MappingAssociation() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        MappingAssociation.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.relationName = null;
                this.srcRootTableName = null;
                this.srcRootDataId = null;
                this.srcParentTableName = null;
                this.srcParentDataId = null;
                this.srcTableName = null;
                this.srcDataId = null;
                this.tarRootTableName = null;
                this.tarRootDataId = null;
                this.tarParentTableName = null;
                this.tarParentDataId = null;
                this.tarTableName = null;
                this.tarDataId = null;
                this.srcTableNamePath = null;
                this.srcDataIdPath = null;
                this.tarTableNamePath = null;
                this.tarDataIdPath = null;
                this.mappingUnitId = null;
                this.associationType = null;
            };
        /**
         * @return {?}
         */
        MappingAssociation.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "relationName", this.relationName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "srcRootTableName", this.srcRootTableName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "srcRootDataId", this.srcRootDataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "srcParentTableName", this.srcParentTableName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "srcParentDataId", this.srcParentDataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "srcTableName", this.srcTableName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "srcDataId", this.srcDataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "tarRootTableName", this.tarRootTableName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "tarRootDataId", this.tarRootDataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "tarParentTableName", this.tarParentTableName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "tarParentDataId", this.tarParentDataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "tarTableName", this.tarTableName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "tarDataId", this.tarDataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "srcTableNamePath", this.srcTableNamePath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "srcDataIdPath", this.srcDataIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "tarTableNamePath", this.tarTableNamePath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "tarDataIdPath", this.tarDataIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingUnitId", this.mappingUnitId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "associationType", this.associationType);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        MappingAssociation.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.relationName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "relationName");
                this.srcRootTableName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcRootTableName");
                this.srcRootDataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcRootDataId");
                this.srcParentTableName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcParentTableName");
                this.srcParentDataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcParentDataId");
                this.srcTableName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcTableName");
                this.srcDataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcDataId");
                this.tarRootTableName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarRootTableName");
                this.tarRootDataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarRootDataId");
                this.tarParentTableName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarParentTableName");
                this.tarParentDataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarParentDataId");
                this.tarTableName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarTableName");
                this.tarDataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarDataId");
                this.srcTableNamePath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcTableNamePath");
                this.srcDataIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcDataIdPath");
                this.tarTableNamePath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarTableNamePath");
                this.tarDataIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarDataIdPath");
                this.mappingUnitId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingUnitId");
                this.associationType = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "associationType");
            };
        return MappingAssociation;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var CreationRuleExecutorAssociations = /** @class */ (function (_super) {
        __extends(CreationRuleExecutorAssociations, _super);
        function CreationRuleExecutorAssociations() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.isAllAssociation = false;
            _this.associations = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        CreationRuleExecutorAssociations.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.creationRuleId = null;
                this.creationRuleInstanceId = null;
                this.bizFlowchartId = null;
                this.creationProcessId = null;
                this.creationProcessInstanceId = null;
                this.isAllAssociation = false;
                this.associations = new Array();
            };
        /**
         * @return {?}
         */
        CreationRuleExecutorAssociations.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessId", this.creationProcessId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessInstanceId", this.creationProcessInstanceId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isAllAssociation", this.isAllAssociation);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        CreationRuleExecutorAssociations.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.creationRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
                this.creationRuleInstanceId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
                this.bizFlowchartId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
                this.creationProcessId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessId");
                this.creationProcessInstanceId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessInstanceId");
                this.isAllAssociation = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isAllAssociation");
                this.associations = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
            };
        return CreationRuleExecutorAssociations;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BefContentVariable = /** @class */ (function (_super) {
        __extends(BefContentVariable, _super);
        function BefContentVariable() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.associations = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        BefContentVariable.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.upperBeFuncSessionId = null;
                this.associations = new Array();
            };
        /**
         * @return {?}
         */
        BefContentVariable.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upperBeFuncSessionId", this.upperBeFuncSessionId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        BefContentVariable.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.upperBeFuncSessionId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upperBeFuncSessionId");
                this.associations = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", CreationRuleExecutorAssociations);
            };
        return BefContentVariable;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MappingResult = /** @class */ (function (_super) {
        __extends(MappingResult, _super);
        function MappingResult() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        MappingResult.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.voId = null;
                this.changeDetail = null;
            };
        /**
         * @return {?}
         */
        MappingResult.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "changeDetail", this.changeDetail);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        MappingResult.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.voId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
                this.changeDetail = JSON.parse(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "changeDetail"));
            };
        return MappingResult;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var CreationMappingResult = /** @class */ (function (_super) {
        __extends(CreationMappingResult, _super);
        function CreationMappingResult() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.associations = new Array();
            _this.entityData = null;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        CreationMappingResult.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.associations = new Array();
                this.entityData = null;
            };
        /**
         * @return {?}
         */
        CreationMappingResult.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "entityData", this.entityData);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        CreationMappingResult.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.associations = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
                this.entityData = commonApi.BpmnModelJsonHelper.GetAttributeValue(obj, "entityData");
            };
        return CreationMappingResult;
    }(MappingResult));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var CreationRuleExecutorArgs = /** @class */ (function (_super) {
        __extends(CreationRuleExecutorArgs, _super);
        function CreationRuleExecutorArgs() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.sourceVoEntities = new Array();
            _this.targetVoProxyEntities = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        CreationRuleExecutorArgs.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.sourceVoEntities = new Array();
                this.creationRuleId = null;
                this.bizFlowchartId = null;
                this.sourceVoId = null;
                this.targetVoProxyId = null;
                this.targetVoProxyEntities = new Array();
            };
        /**
         * @return {?}
         */
        CreationRuleExecutorArgs.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntities", this.sourceVoEntities);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyId", this.targetVoProxyId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyEntities", this.targetVoProxyEntities);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        CreationRuleExecutorArgs.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.creationRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
                this.sourceVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
                this.bizFlowchartId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
                this.sourceVoEntities = commonApi.BpmnModelJsonHelper.GetAttributeValue(obj, "sourceVoEntities");
                this.targetVoProxyId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyId");
                this.targetVoProxyEntities = commonApi.BpmnModelJsonHelper.GetAttributeValue(obj, "targetVoProxyEntities");
            };
        return CreationRuleExecutorArgs;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var CreationRuleExecutorResult = /** @class */ (function (_super) {
        __extends(CreationRuleExecutorResult, _super);
        function CreationRuleExecutorResult() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.isAllAssociation = false;
            _this.mappingResults = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        CreationRuleExecutorResult.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.creationRuleId = null;
                this.creationRuleInstanceId = null;
                this.bizFlowchartId = null;
                this.isAllAssociation = false;
                this.mappingResults = new Array();
            };
        /**
         * @return {?}
         */
        CreationRuleExecutorResult.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isAllAssociation", this.isAllAssociation);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingResults", this.mappingResults);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        CreationRuleExecutorResult.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.creationRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
                this.creationRuleInstanceId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
                this.bizFlowchartId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
                this.isAllAssociation = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isAllAssociation");
                this.mappingResults = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingResults", CreationMappingResult);
            };
        return CreationRuleExecutorResult;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ConditionResult = /** @class */ (function (_super) {
        __extends(ConditionResult, _super);
        function ConditionResult() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        ConditionResult.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.conditionExecutionResult = false;
                this.message = null;
            };
        /**
         * @return {?}
         */
        ConditionResult.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "conditionExecutionResult", this.conditionExecutionResult);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "message", this.message);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        ConditionResult.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.conditionExecutionResult = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "conditionExecutionResult");
                this.message = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "message");
            };
        return ConditionResult;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RpcVoChangeDetail = /** @class */ (function (_super) {
        __extends(RpcVoChangeDetail, _super);
        function RpcVoChangeDetail() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        RpcVoChangeDetail.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.voId = null;
                this.changeDetail = null;
            };
        /**
         * @return {?}
         */
        RpcVoChangeDetail.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "changeDetail", this.changeDetail);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RpcVoChangeDetail.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.voId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
                this.changeDetail = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "changeDetail");
            };
        return RpcVoChangeDetail;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * RPC映射结果。
     * @abstract
     */
    var /**
     * RPC映射结果。
     * @abstract
     */ RpcMappingResult = /** @class */ (function (_super) {
        __extends(RpcMappingResult, _super);
        function RpcMappingResult() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return RpcMappingResult;
    }(RpcVoChangeDetail));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RpcCreationMappingResult = /** @class */ (function (_super) {
        __extends(RpcCreationMappingResult, _super);
        function RpcCreationMappingResult() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.associations = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        RpcCreationMappingResult.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.associations = new Array();
            };
        /**
         * @return {?}
         */
        RpcCreationMappingResult.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RpcCreationMappingResult.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.associations = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
            };
        return RpcCreationMappingResult;
    }(RpcMappingResult));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RpcCreationRuleExecutorResult = /** @class */ (function (_super) {
        __extends(RpcCreationRuleExecutorResult, _super);
        function RpcCreationRuleExecutorResult() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.mappingResults = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        RpcCreationRuleExecutorResult.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.creationRuleId = null;
                this.creationRuleInstanceId = null;
                this.bizFlowchartId = null;
                this.mappingResults = new Array();
            };
        /**
         * @return {?}
         */
        RpcCreationRuleExecutorResult.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingResults", this.mappingResults);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RpcCreationRuleExecutorResult.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.creationRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
                this.creationRuleInstanceId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
                this.bizFlowchartId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
                this.mappingResults = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingResults", RpcCreationMappingResult);
            };
        return RpcCreationRuleExecutorResult;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormBillCreationProcessExecutorResult = /** @class */ (function (_super) {
        __extends(FormBillCreationProcessExecutorResult, _super);
        function FormBillCreationProcessExecutorResult() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.pushCreationMode = commonApi.PushCreationMode.CREATION;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        FormBillCreationProcessExecutorResult.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.conditionResult = null;
                this.creationRuleExecutorResult = null;
                this.pushCreationMode = commonApi.PushCreationMode.CREATION;
                this.downBillSuCode = null;
            };
        /**
         * @return {?}
         */
        FormBillCreationProcessExecutorResult.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "conditionResult", this.conditionResult);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleExecutorResult", this.creationRuleExecutorResult);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "pushCreationMode", this.pushCreationMode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillSuCode", this.downBillSuCode);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        FormBillCreationProcessExecutorResult.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.conditionResult = commonApi.BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "conditionResult", ConditionResult);
                this.creationRuleExecutorResult = commonApi.BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "creationRuleExecutorResult", RpcCreationRuleExecutorResult);
                /** @type {?} */
                var pushCreationMode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "pushCreationMode");
                if (pushCreationMode != null) {
                    this.pushCreationMode = commonApi.PushCreationMode[pushCreationMode];
                }
                this.downBillSuCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillSuCode");
            };
        return FormBillCreationProcessExecutorResult;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RpcCreationRuleExecutorArgs = /** @class */ (function (_super) {
        __extends(RpcCreationRuleExecutorArgs, _super);
        function RpcCreationRuleExecutorArgs() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.sourceVoEntities = new Array();
            _this.targetVoProxyEntities = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        RpcCreationRuleExecutorArgs.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.sourceVoEntities = new Array();
                this.creationRuleId = null;
                this.bizFlowchartId = null;
                this.sourceVoId = null;
                this.targetVoProxyId = null;
                this.targetVoProxyEntities = new Array();
            };
        /**
         * @return {?}
         */
        RpcCreationRuleExecutorArgs.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntities", commonApi.BpmnModelJsonHelper.StringListToJson(this.sourceVoEntities));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyId", this.targetVoProxyId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyEntities", commonApi.BpmnModelJsonHelper.StringListToJson(this.targetVoProxyEntities));
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RpcCreationRuleExecutorArgs.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.creationRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
                this.sourceVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
                this.bizFlowchartId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
                this.sourceVoEntities = commonApi.BpmnModelJsonHelper.StringListFromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoEntities"));
                this.targetVoProxyId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyId");
                this.targetVoProxyEntities = commonApi.BpmnModelJsonHelper.StringListFromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyEntities"));
            };
        return RpcCreationRuleExecutorArgs;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RpcFireBizEventArgs = /** @class */ (function (_super) {
        __extends(RpcFireBizEventArgs, _super);
        function RpcFireBizEventArgs() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.voChangeDetails = null;
            _this.voEntityDatas = null;
            _this.voEntityDataIds = new Array();
            _this.associations = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        RpcFireBizEventArgs.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.voId = null;
                this.voChangeDetails = null;
                this.voEntityDatas = null;
                this.voEntityDataIds = new Array();
                this.bizEventId = null;
                this.voFuncSessionId = null;
                this.bizFlowchartId = null;
                this.upperBeFuncSessionId = null;
                this.associations = new Array();
            };
        /**
         * @return {?}
         */
        RpcFireBizEventArgs.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "voChangeDetails", this.voChangeDetails);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "voEntityDatas", this.voEntityDatas);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "voEntityDataIds", commonApi.BpmnModelJsonHelper.StringListToJsonArray(this.voEntityDataIds));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventId", this.bizEventId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "voFuncSessionId", this.voFuncSessionId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upperBeFuncSessionId", this.upperBeFuncSessionId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RpcFireBizEventArgs.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.voId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
                this.voChangeDetails = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voChangeDetails");
                this.voEntityDatas = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voEntityDatas");
                this.voEntityDataIds = commonApi.BpmnModelJsonHelper.StringListFromJsonArray(commonApi.BpmnModelJsonHelper.GetAttributeValue(obj, "voEntityDataIds"));
                this.bizEventId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventId");
                this.voFuncSessionId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voFuncSessionId");
                this.bizFlowchartId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
                this.upperBeFuncSessionId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upperBeFuncSessionId");
                this.associations = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", CreationRuleExecutorAssociations);
            };
        return RpcFireBizEventArgs;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RpcWritebackMappingResult = /** @class */ (function (_super) {
        __extends(RpcWritebackMappingResult, _super);
        function RpcWritebackMappingResult() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return RpcWritebackMappingResult;
    }(RpcMappingResult));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RpcWritebackRuleExecutorArgs = /** @class */ (function (_super) {
        __extends(RpcWritebackRuleExecutorArgs, _super);
        function RpcWritebackRuleExecutorArgs() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.sourceVoEntities = new Array();
            _this.associations = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        RpcWritebackRuleExecutorArgs.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.writebackRuleId = null;
                this.sourceVoId = null;
                this.targetVoFuncSessionId = null;
                this.sourceVoEntities = new Array();
                this.associations = new Array();
            };
        /**
         * @return {?}
         */
        RpcWritebackRuleExecutorArgs.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "writebackRuleId", this.writebackRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoFuncSessionId", this.targetVoFuncSessionId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntities", commonApi.BpmnModelJsonHelper.StringListToJson(this.sourceVoEntities));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RpcWritebackRuleExecutorArgs.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.writebackRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackRuleId");
                this.sourceVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
                this.targetVoFuncSessionId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoFuncSessionId");
                this.sourceVoEntities = commonApi.BpmnModelJsonHelper.StringListFromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoEntities"));
                this.associations = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
            };
        return RpcWritebackRuleExecutorArgs;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RpcWritebackRuleExecutorResult = /** @class */ (function (_super) {
        __extends(RpcWritebackRuleExecutorResult, _super);
        function RpcWritebackRuleExecutorResult() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.mappingResults = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        RpcWritebackRuleExecutorResult.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.mappingResults = new Array();
            };
        /**
         * @return {?}
         */
        RpcWritebackRuleExecutorResult.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingResults", this.mappingResults);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RpcWritebackRuleExecutorResult.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.mappingResults = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingResults", RpcWritebackMappingResult);
            };
        return RpcWritebackRuleExecutorResult;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var AifManualCreationType = {
        Push: 'Push',
        Refer: 'Refer',
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @deprecated please use \@edp-bif/runtime-api
     */
    var /**
     * @deprecated please use \@edp-bif/runtime-api
     */ AifRuntimeModuleConfig = /** @class */ (function () {
        function AifRuntimeModuleConfig() {
        }
        return AifRuntimeModuleConfig;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @deprecated please use \@edp-bif/runtime-api
     * @type {?}
     */
    var SELECTOR_ARGS_INJECTOR = new core.InjectionToken("SELECTOR_ARGS_INJECTOR");

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @abstract
     */
    var /**
     * @abstract
     */ FormRestArgs = /** @class */ (function (_super) {
        __extends(FormRestArgs, _super);
        function FormRestArgs() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        FormRestArgs.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.rpcCreationRuleExecutorResult = null;
            };
        /**
         * @return {?}
         */
        FormRestArgs.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "rpcCreationRuleExecutorResult", this.rpcCreationRuleExecutorResult);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        FormRestArgs.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.rpcCreationRuleExecutorResult = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "rpcCreationRuleExecutorResult");
            };
        return FormRestArgs;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormCreateArgs = /** @class */ (function (_super) {
        __extends(FormCreateArgs, _super);
        function FormCreateArgs() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        FormCreateArgs.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.formVoId = null;
            };
        /**
         * @return {?}
         */
        FormCreateArgs.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "formVoId", this.formVoId);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        FormCreateArgs.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.formVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formVoId");
            };
        return FormCreateArgs;
    }(FormRestArgs));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @abstract
     */
    var /**
     * @abstract
     */ FormRestResult = /** @class */ (function (_super) {
        __extends(FormRestResult, _super);
        function FormRestResult() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return FormRestResult;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormCreateResult = /** @class */ (function (_super) {
        __extends(FormCreateResult, _super);
        function FormCreateResult() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.entityDataList = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        FormCreateResult.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.entityDataList = new Array();
            };
        /**
         * @return {?}
         */
        FormCreateResult.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "entityDataList", commonApi.BpmnModelJsonHelper.StringListToJsonArray(this.entityDataList));
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        FormCreateResult.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.entityDataList = commonApi.BpmnModelJsonHelper.StringListFromJsonArray(commonApi.BpmnModelJsonHelper.GetAttributeValue(obj, "entityDataList"));
            };
        return FormCreateResult;
    }(FormRestResult));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 表单手工生单时执行生单规则参数类。
     */
    var /**
     * 表单手工生单时执行生单规则参数类。
     */ FormCreationRuleExecutorArgs = /** @class */ (function (_super) {
        __extends(FormCreationRuleExecutorArgs, _super);
        function FormCreationRuleExecutorArgs() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            /**
             * 来源单据ID树形列表
             */
            _this.sourceVoEntityDatas = new Array();
            /**
             * 目标单据主表ID列表
             */
            _this.targetVoDataIds = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        FormCreationRuleExecutorArgs.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.sourceVoEntityDatas = new Array();
                this.creationRuleId = null;
                this.bizFlowchartId = null;
                this.targetVoDataIds = new Array();
            };
        /**
         * @return {?}
         */
        FormCreationRuleExecutorArgs.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntityDatas", commonApi.BpmnModelJsonHelper.StringListToJsonArray(this.sourceVoEntityDatas));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoDataIds", commonApi.BpmnModelJsonHelper.StringListToJsonArray(this.targetVoDataIds));
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        FormCreationRuleExecutorArgs.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.creationRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
                this.bizFlowchartId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
                this.sourceVoEntityDatas = commonApi.BpmnModelJsonHelper.StringListFromJsonArray(commonApi.BpmnModelJsonHelper.GetAttributeValue(obj, "sourceVoEntityDatas"));
                this.targetVoDataIds = commonApi.BpmnModelJsonHelper.StringListFromJsonArray(( /** @type {?} */(commonApi.BpmnModelJsonHelper.GetAttributeValue(obj, "targetVoDataIds"))));
            };
        return FormCreationRuleExecutorArgs;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormEntityData = /** @class */ (function (_super) {
        __extends(FormEntityData, _super);
        function FormEntityData() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.children = {};
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        FormEntityData.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.dataId = null;
                if (this.children != null) {
                    for (var key in this.children) {
                        delete this.children[key];
                    }
                }
            };
        /**
         * @return {?}
         */
        FormEntityData.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "dataId", this.dataId);
                if (this.children != null) {
                    /** @type {?} */
                    var childrenObj = {};
                    for (var key in this.children) {
                        commonApi.BpmnModelJsonHelper.SetAttributeValue(childrenObj, key, this.children[key]);
                    }
                    commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "children", childrenObj);
                }
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        FormEntityData.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.dataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dataId");
                /** @type {?} */
                var childrenObj = commonApi.BpmnModelJsonHelper.GetAttributeValue(obj, "children");
                if (childrenObj != null) {
                    for (var key in childrenObj) {
                        /** @type {?} */
                        var value = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(childrenObj, key, FormEntityData);
                        this.children[key] = value;
                    }
                }
            };
        return FormEntityData;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormUpdateArgs = /** @class */ (function (_super) {
        __extends(FormUpdateArgs, _super);
        function FormUpdateArgs() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return FormUpdateArgs;
    }(FormRestArgs));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormUpdateResult = /** @class */ (function (_super) {
        __extends(FormUpdateResult, _super);
        function FormUpdateResult() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return FormUpdateResult;
    }(FormRestResult));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormGetCreationRuleEntityExArgs = /** @class */ (function (_super) {
        __extends(FormGetCreationRuleEntityExArgs, _super);
        function FormGetCreationRuleEntityExArgs() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        FormGetCreationRuleEntityExArgs.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.bizFlowchartId = null;
                this.sourceVoId = null;
                this.targetVoId = null;
            };
        /**
         * @return {?}
         */
        FormGetCreationRuleEntityExArgs.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoId", this.targetVoId);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        FormGetCreationRuleEntityExArgs.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.bizFlowchartId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
                this.sourceVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
                this.targetVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoId");
            };
        return FormGetCreationRuleEntityExArgs;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormGetCreationRuleEntityExResult = /** @class */ (function (_super) {
        __extends(FormGetCreationRuleEntityExResult, _super);
        function FormGetCreationRuleEntityExResult() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.ruleEntityExes = new Array();
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        FormGetCreationRuleEntityExResult.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.ruleEntityExes = new Array();
            };
        /**
         * @return {?}
         */
        FormGetCreationRuleEntityExResult.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "ruleEntityExes", this.ruleEntityExes);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        FormGetCreationRuleEntityExResult.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.ruleEntityExes = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(obj, "ruleEntityExes", commonApi.SlimCreationRuleEntityEx);
            };
        return FormGetCreationRuleEntityExResult;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var AifDataService = /** @class */ (function () {
        function AifDataService(repository) {
            this.getCreationRuleEntityByIdUrl = "/api/runtime/aif/v1.0/form/getcreationruleentityex";
            this.getCreationRuleEntityByIdMethod = "GET";
            this.getCreationRuleUrl = "/api/runtime/aif/v1.0/form/getcreationruleentityexes";
            this.getCreationRuleMethod = "POST";
            this.executeCreationRuleUrl = "/api/runtime/aif/v1.0/form/executecreationrule";
            this.executeCreationRuleMethod = "POST";
            this.formCreateUrl = "/api/runtime/aif/v1.0/form/create";
            this.formCreateMethod = "POST";
            this.formUpdateUrl = "/api/runtime/aif/v1.0/form/update";
            this.formUpdateMethod = "POST";
            this.generalTextUrl = "/api/runtime/aif/v1.0/generaltext";
            this.generalTextMethod = "GET";
            this.repository = ( /** @type {?} */(repository));
        }
        /**
         * 根据来源 VO Id获取生单规则实体列表。
         *
         * @param sourceVoId 来源 VO Id。
         * @return 符合条件的生单规则实体列表。
         */
        /**
         * 根据来源 VO Id获取生单规则实体列表。
         *
         * @param {?} sourceVoId 来源 VO Id。
         * @return {?} 符合条件的生单规则实体列表。
         */
        AifDataService.prototype.getSlimCreationRulesBySourceVoId = /**
         * 根据来源 VO Id获取生单规则实体列表。
         *
         * @param {?} sourceVoId 来源 VO Id。
         * @return {?} 符合条件的生单规则实体列表。
         */
            function (sourceVoId) {
                /** @type {?} */
                var args = new FormGetCreationRuleEntityExArgs();
                args.sourceVoId = sourceVoId;
                return this.getCreationRuleList(args).pipe(operators.map(( /**
                 * @param {?} res
                 * @return {?}
                 */function (res) {
                    return res.ruleEntityExes;
                })));
            };
        /**
         * 根据目标 VO Id获取生单规则实体列表。
         *
         * @param targetVoId 目标 VO Id。
         * @return 符合条件的生单规则实体列表。
         */
        /**
         * 根据目标 VO Id获取生单规则实体列表。
         *
         * @param {?} targetVoId 目标 VO Id。
         * @return {?} 符合条件的生单规则实体列表。
         */
        AifDataService.prototype.getSlimCreationRulesByTargetVoId = /**
         * 根据目标 VO Id获取生单规则实体列表。
         *
         * @param {?} targetVoId 目标 VO Id。
         * @return {?} 符合条件的生单规则实体列表。
         */
            function (targetVoId) {
                /** @type {?} */
                var args = new FormGetCreationRuleEntityExArgs();
                args.targetVoId = targetVoId;
                return this.getCreationRuleList(args).pipe(operators.map(( /**
                 * @param {?} res
                 * @return {?}
                 */function (res) {
                    return res.ruleEntityExes;
                })));
            };
        /**
         * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
         *
         * @param sourceVoId 来源 VO Id。
         * @param targetVoId 目标 VO Id。
         * @return 符合条件的生单规则实体列表。
         */
        /**
         * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
         *
         * @param {?} sourceVoId 来源 VO Id。
         * @param {?} targetVoId 目标 VO Id。
         * @return {?} 符合条件的生单规则实体列表。
         */
        AifDataService.prototype.getSlimCreationRulesBySourceVoIdAndTargetVoId = /**
         * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
         *
         * @param {?} sourceVoId 来源 VO Id。
         * @param {?} targetVoId 目标 VO Id。
         * @return {?} 符合条件的生单规则实体列表。
         */
            function (sourceVoId, targetVoId) {
                /** @type {?} */
                var args = new FormGetCreationRuleEntityExArgs();
                args.sourceVoId = sourceVoId;
                args.targetVoId = targetVoId;
                return this.getCreationRuleList(args).pipe(operators.map(( /**
                 * @param {?} res
                 * @return {?}
                 */function (res) {
                    return res.ruleEntityExes;
                })));
            };
        /**
         * 根据来源 VO Id获取生单规则实体列表。
         *
         * @param sourceVoId     来源 VO Id。
         * @param bizFlowchartId 业务场景图Id。
         * @return 符合条件的生单规则实体列表。
         */
        /**
         * 根据来源 VO Id获取生单规则实体列表。
         *
         * @param {?} sourceVoId     来源 VO Id。
         * @param {?} bizFlowchartId 业务场景图Id。
         * @return {?} 符合条件的生单规则实体列表。
         */
        AifDataService.prototype.getSlimCreationRulesBySourceVoIdAndFlowchartId = /**
         * 根据来源 VO Id获取生单规则实体列表。
         *
         * @param {?} sourceVoId     来源 VO Id。
         * @param {?} bizFlowchartId 业务场景图Id。
         * @return {?} 符合条件的生单规则实体列表。
         */
            function (sourceVoId, bizFlowchartId) {
                /** @type {?} */
                var args = new FormGetCreationRuleEntityExArgs();
                args.sourceVoId = sourceVoId;
                args.bizFlowchartId = bizFlowchartId;
                return this.getCreationRuleList(args).pipe(operators.map(( /**
                 * @param {?} res
                 * @return {?}
                 */function (res) {
                    return res.ruleEntityExes;
                })));
            };
        /**
         * 根据目标 VO Id获取生单规则实体列表。
         *
         * @param targetVoId     目标 VO Id。
         * @param bizFlowchartId 业务场景图Id。
         * @return 符合条件的生单规则实体列表。
         */
        /**
         * 根据目标 VO Id获取生单规则实体列表。
         *
         * @param {?} targetVoId     目标 VO Id。
         * @param {?} bizFlowchartId 业务场景图Id。
         * @return {?} 符合条件的生单规则实体列表。
         */
        AifDataService.prototype.getSlimCreationRulesByTargetVoIdAndFlowchartId = /**
         * 根据目标 VO Id获取生单规则实体列表。
         *
         * @param {?} targetVoId     目标 VO Id。
         * @param {?} bizFlowchartId 业务场景图Id。
         * @return {?} 符合条件的生单规则实体列表。
         */
            function (targetVoId, bizFlowchartId) {
                /** @type {?} */
                var args = new FormGetCreationRuleEntityExArgs();
                args.targetVoId = targetVoId;
                args.bizFlowchartId = bizFlowchartId;
                return this.getCreationRuleList(args).pipe(operators.map(( /**
                 * @param {?} res
                 * @return {?}
                 */function (res) {
                    return res.ruleEntityExes;
                })));
            };
        /**
         * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
         *
         * @param sourceVoId     来源 VO Id。
         * @param targetVoId     目标 VO Id。
         * @param bizFlowchartId 业务场景图Id。
         * @return 符合条件的生单规则实体列表。
         */
        /**
         * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
         *
         * @param {?} sourceVoId     来源 VO Id。
         * @param {?} targetVoId     目标 VO Id。
         * @param {?} bizFlowchartId 业务场景图Id。
         * @return {?} 符合条件的生单规则实体列表。
         */
        AifDataService.prototype.getSlimCreationRulesBySourceVoIdAndTargetVoIdAndFlowchartId = /**
         * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
         *
         * @param {?} sourceVoId     来源 VO Id。
         * @param {?} targetVoId     目标 VO Id。
         * @param {?} bizFlowchartId 业务场景图Id。
         * @return {?} 符合条件的生单规则实体列表。
         */
            function (sourceVoId, targetVoId, bizFlowchartId) {
                /** @type {?} */
                var args = new FormGetCreationRuleEntityExArgs();
                args.sourceVoId = sourceVoId;
                args.targetVoId = targetVoId;
                args.bizFlowchartId = bizFlowchartId;
                return this.getCreationRuleList(args).pipe(operators.map(( /**
                 * @param {?} res
                 * @return {?}
                 */function (res) {
                    return res.ruleEntityExes;
                })));
            };
        /**
         * 获取生单规则实体
         */
        /**
         * 获取生单规则实体
         * @param {?} creationRuleId
         * @return {?}
         */
        AifDataService.prototype.getSlimCreationRuleEntityExById = /**
         * 获取生单规则实体
         * @param {?} creationRuleId
         * @return {?}
         */
            function (creationRuleId) {
                /** @type {?} */
                var getCreationRuleById$ = this.repository.proxy.request(this.getCreationRuleEntityByIdUrl + "/" + creationRuleId, this.getCreationRuleEntityByIdMethod, null);
                return getCreationRuleById$;
            };
        /**
         * 获取表单配置实体
         * @param formSettingsId
         */
        /**
         * 获取表单配置实体
         * @param {?} formSettingsId
         * @return {?}
         */
        AifDataService.prototype.getFormSettingsById = /**
         * 获取表单配置实体
         * @param {?} formSettingsId
         * @return {?}
         */
            function (formSettingsId) {
                if (formSettingsId == null || formSettingsId.trim() == "") {
                    throw "生单规则为空";
                }
                /** @type {?} */
                var formSettings$ = this.repository.proxy.request(this.generalTextUrl + "/" + formSettingsId, this.generalTextMethod, null, { responseType: "text" });
                return formSettings$.pipe(operators.map(( /**
                 * @param {?} data
                 * @return {?}
                 */function (data) {
                    /** @type {?} */
                    var generateText = new commonApi.GeneralTextEntity();
                    generateText.LoadFromJson(data);
                    /** @type {?} */
                    var formSettings = new commonApi.FormSettings();
                    formSettings.LoadFromJson(generateText.text);
                    return formSettings;
                })));
            };
        /**
         * 执行生单规则
         */
        /**
         * 执行生单规则
         * @param {?} args
         * @return {?}
         */
        AifDataService.prototype.executeCreationRule = /**
         * 执行生单规则
         * @param {?} args
         * @return {?}
         */
            function (args) {
                /** @type {?} */
                var body = args.ConvertToJson();
                /** @type {?} */
                var executeCreationRule$ = this.repository.proxy.request(this.executeCreationRuleUrl, this.executeCreationRuleMethod, null, {
                    body: body,
                    responseType: 'text'
                });
                return executeCreationRule$.pipe(operators.map(( /**
                 * @param {?} data
                 * @return {?}
                 */function (data) {
                    /** @type {?} */
                    var res = new RpcCreationRuleExecutorResult();
                    res.LoadFromJson(data);
                    return res;
                })));
            };
        /**
         * 创建单据
         */
        /**
         * 创建单据
         * @param {?} args
         * @return {?}
         */
        AifDataService.prototype.createEntityDatas = /**
         * 创建单据
         * @param {?} args
         * @return {?}
         */
            function (args) {
                var _this = this;
                /** @type {?} */
                var body = args.ConvertToJson();
                /** @type {?} */
                var options = {
                    body: body,
                    responseType: 'text'
                };
                /** @type {?} */
                var create$ = this.repository.proxy.request(this.formCreateUrl, this.formCreateMethod, null, options);
                return create$.pipe(operators.switchMap(( /**
                 * @param {?} res
                 * @return {?}
                 */function (res) {
                    /** @type {?} */
                    var result = new FormCreateResult();
                    result.LoadFromJson(res);
                    /** @type {?} */
                    var entities = result.entityDataList.map(( /**
                     * @param {?} entityStr
                     * @return {?}
                     */function (entityStr) {
                        /** @type {?} */
                        var newEntity = _this.repository.buildEntity(JSON.parse(entityStr));
                        _this.appendInitialData(newEntity, newEntity);
                        return newEntity;
                    }));
                    /** @type {?} */
                    var origin = _this.repository.entityCollection.getAllEntities();
                    if (origin == null || origin.length == 0) {
                        _this.repository.entityCollection.loadEntities(entities);
                    }
                    else {
                        _this.repository.entityCollection.addEntities(entities);
                    }
                    return rxjs.of(result);
                })));
            };
        /**
         * 修改单据
         */
        /**
         * 修改单据
         * @param {?} args
         * @return {?}
         */
        AifDataService.prototype.updateEntityDatas = /**
         * 修改单据
         * @param {?} args
         * @return {?}
         */
            function (args) {
                /** @type {?} */
                var body = args.ConvertToJson();
                /** @type {?} */
                var options = {
                    body: body,
                    responseType: "text"
                };
                /** @type {?} */
                var update$ = this.repository.proxy.request(this.formUpdateUrl, this.formUpdateMethod, null, options);
                return update$.pipe(operators.map(( /**
                 * @param {?} data
                 * @return {?}
                 */function (data) {
                    /** @type {?} */
                    var res = new FormUpdateResult();
                    res.LoadFromJson(data);
                    return res;
                })));
            };
        /**
         * @private
         * @param {?} args
         * @return {?}
         */
        AifDataService.prototype.getCreationRuleList = /**
         * @private
         * @param {?} args
         * @return {?}
         */
            function (args) {
                if (args == null) {
                    throw "args不可为空";
                }
                if (typeof args == "string") {
                    if (args == "") {
                        throw "args不可为空";
                    }
                    /** @type {?} */
                    var temp = new FormGetCreationRuleEntityExArgs();
                    temp.LoadFromJson(args);
                    args = temp;
                }
                /** @type {?} */
                var getCreationRule$ = this.repository.proxy.request(this.getCreationRuleUrl, this.getCreationRuleMethod, null, { body: args.ConvertToJson(), responseType: "text" });
                return getCreationRule$.pipe(operators.map(( /**
                 * @param {?} data
                 * @return {?}
                 */function (data) {
                    /** @type {?} */
                    var result = new FormGetCreationRuleEntityExResult();
                    result.LoadFromJson(data);
                    return result;
                })));
            };
        /**
     * 参考farris/bef EntityUtil实现
     * @param entity
     * @param initialData
     */
        /**
         * 参考farris/bef EntityUtil实现
         * @private
         * @param {?} entity
         * @param {?} initialData
         * @return {?}
         */
        AifDataService.prototype.appendInitialData = /**
         * 参考farris/bef EntityUtil实现
         * @private
         * @param {?} entity
         * @param {?} initialData
         * @return {?}
         */
            function (entity, initialData) {
                /** @type {?} */
                var data = Object.assign({}, initialData);
                delete data.id;
                delete data.parentID;
                entity['initialData'] = data;
            };
        AifDataService.decorators = [
            { type: core.Injectable }
        ];
        /** @nocollapse */
        AifDataService.ctorParameters = function () {
            return [
                { type: devkit.Repository }
            ];
        };
        return AifDataService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @deprecated please use \@edp-bif/runtime-api
     */
    var ReferSelectorComponent = /** @class */ (function () {
        function ReferSelectorComponent(args, dataService, localeId) {
            var _this = this;
            this.dataService = dataService;
            //来源VoType列表
            this.sourceVoIdList = [];
            //来源VoType与来源VoName映射
            this.sourceVoNameDict = {};
            //来源VoType与相应的生单规则列表映射
            this.creationRuleDict = {};
            this.bsModalRef4AifSelector = null;
            this.subject4CreationRule = new rxjs.Subject();
            this.clickOk4AifSelector = ( /**
             * @return {?}
             */function () {
                _this.subject4CreationRule.next(_this.selectedCreationRule);
                _this.subject4CreationRule.complete();
                _this.bsModalRef4AifSelector.close();
            });
            this.clickClose4AifSelector = ( /**
             * @return {?}
             */function () {
                _this.bsModalRef4AifSelector.close();
            });
            this.args = new FormGetCreationRuleEntityExArgs();
            this.args.LoadFromJsonObject(args);
        }
        Object.defineProperty(ReferSelectorComponent.prototype, "selectedSourceVoId", {
            get: /**
             * @return {?}
             */ function () {
                return this._selectedSourceVoId;
            },
            set: /**
             * @param {?} sourceVoId
             * @return {?}
             */ function (sourceVoId) {
                this._selectedSourceVoId = sourceVoId;
                if (sourceVoId != null) {
                    this.selectedCreationRule = this.creationRuleDict[this._selectedSourceVoId][0];
                }
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        ReferSelectorComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.loadItems();
            };
        /**
         * @private
         * @return {?}
         */
        ReferSelectorComponent.prototype.loadItems = /**
         * @private
         * @return {?}
         */
            function () {
                var _this = this;
                /** @type {?} */
                var list$ = null;
                if (this.args.sourceVoId == null) {
                    if (this.args.bizFlowchartId == null) {
                        list$ = this.dataService.getSlimCreationRulesByTargetVoId(this.args.targetVoId);
                    }
                    else {
                        list$ = this.dataService.getSlimCreationRulesByTargetVoIdAndFlowchartId(this.args.targetVoId, this.args.bizFlowchartId);
                    }
                }
                else {
                    if (this.args.bizFlowchartId == null) {
                        list$ = this.dataService.getSlimCreationRulesBySourceVoIdAndTargetVoId(this.args.sourceVoId, this.args.targetVoId);
                    }
                    else {
                        list$ = this.dataService.getSlimCreationRulesBySourceVoIdAndTargetVoIdAndFlowchartId(this.args.sourceVoId, this.args.targetVoId, this.args.bizFlowchartId);
                    }
                }
                list$.subscribe(( /**
                 * @param {?} result
                 * @return {?}
                 */function (result) {
                    /** @type {?} */
                    var list = result;
                    list.forEach(( /**
                     * @param {?} elt
                     * @return {?}
                     */function (elt) {
                        if (_this.sourceVoIdList.includes(elt.sourceVoId) === false) {
                            _this.sourceVoIdList.push(elt.sourceVoId);
                            _this.creationRuleDict[elt.sourceVoId] = new Array();
                        }
                        _this.sourceVoNameDict[elt.sourceVoId] = elt.sourceVoName;
                        _this.creationRuleDict[elt.sourceVoId].push(elt);
                    }));
                    _this.selectedSourceVoId = _this.sourceVoIdList.length > 0 ? _this.sourceVoIdList[0] : null;
                }));
            };
        ReferSelectorComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'lib-refer-selector',
                        template: "<div class=\"g-main p-2\">\r\n  <!--  -->\r\n  <div class=\"form-group\" style=\"height: 150px;overflow-y: auto;\">\r\n    <div [hidden]=\"sourceVoIdList.length != 0\"><span style=\"color: red;\">{{ 'emptyResult' | aifLocale }}</span></div>\r\n    <div>{{ 'selectSourceBillType' | aifLocale }}</div>\r\n    <div *ngFor=\"let sourceVoId of sourceVoIdList;let i=index\" class=\"col-sm-12 col-md-5 form-check form-check-inline\">\r\n      <input class=\"form-check-input\" type=\"radio\" [name]=\"'sourceVoId'+i\" id=\"sourceVoId\" [value]=\"sourceVoId\"\r\n        [(ngModel)]=\"selectedSourceVoId\" />\r\n      <label class=\"form-check-label\" [for]=\"'sourceVoId'+i\">{{sourceVoNameDict[sourceVoId]}}</label>\r\n    </div>\r\n  </div>\r\n\r\n  <!--  -->\r\n  <div class=\"form-group\" style=\"\">\r\n    <div>{{ 'selectCreationRule' | aifLocale }}</div>\r\n    <select class=\"form-control\" name=\"selectedCreationRule\" [(ngModel)]=\"selectedCreationRule\">\r\n      <option *ngFor=\"let cr of creationRuleDict[selectedSourceVoId];\" [ngValue]=\"cr\">\r\n        {{cr.name}}</option>\r\n    </select>\r\n  </div>\r\n\r\n  <!--  -->\r\n  <div class=\"form-group\" style=\"\">\r\n    <div>{{ 'creationRuleDesciption' | aifLocale }}</div>\r\n    <textarea readonly class=\"form-control\" name=\"description\" rows=\"3\"\r\n      [ngModel]=\"selectedCreationRule&&selectedCreationRule.description\"></textarea>\r\n  </div>\r\n</div>",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        ReferSelectorComponent.ctorParameters = function () {
            return [
                { type: undefined, decorators: [{ type: core.Inject, args: [SELECTOR_ARGS_INJECTOR,] }] },
                { type: AifDataService },
                { type: String, decorators: [{ type: core.Inject, args: [core.LOCALE_ID,] }] }
            ];
        };
        return ReferSelectorComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @deprecated please use \@edp-bif/runtime-api
     */
    var PushSelectorComponent = /** @class */ (function () {
        function PushSelectorComponent(args, dataService, localeId) {
            var _this = this;
            this.dataService = dataService;
            //目标VoId列表
            this.targetVoIdList = [];
            //目标VoId与目标VoName映射
            this.targetVoNameDict = {};
            //目标VoId与相应的生单规则列表映射
            this.creationRuleDict = {};
            this.bsModalRef4AifSelector = null;
            this.subject4CreationRule = new rxjs.Subject();
            this.clickOk4AifSelector = ( /**
             * @return {?}
             */function () {
                _this.subject4CreationRule.next(_this.selectedCreationRule);
                _this.subject4CreationRule.complete();
                _this.bsModalRef4AifSelector.close();
            });
            this.clickClose4AifSelector = ( /**
             * @return {?}
             */function () {
                _this.bsModalRef4AifSelector.close();
            });
            this.args = new FormGetCreationRuleEntityExArgs();
            this.args.LoadFromJsonObject(args);
        }
        Object.defineProperty(PushSelectorComponent.prototype, "selectedTargetVoId", {
            get: /**
             * @return {?}
             */ function () {
                return this._selectedTargetVoId;
            },
            set: /**
             * @param {?} targetVoId
             * @return {?}
             */ function (targetVoId) {
                this._selectedTargetVoId = targetVoId;
                if (targetVoId != null) {
                    this.selectedCreationRule = this.creationRuleDict[this._selectedTargetVoId][0];
                }
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        PushSelectorComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.loadItems();
            };
        /**
         * @private
         * @return {?}
         */
        PushSelectorComponent.prototype.loadItems = /**
         * @private
         * @return {?}
         */
            function () {
                var _this = this;
                /** @type {?} */
                var list$ = null;
                if (this.args.targetVoId == null) {
                    if (this.args.bizFlowchartId == null) {
                        list$ = this.dataService.getSlimCreationRulesBySourceVoId(this.args.sourceVoId);
                    }
                    else {
                        list$ = this.dataService.getSlimCreationRulesBySourceVoIdAndFlowchartId(this.args.sourceVoId, this.args.bizFlowchartId);
                    }
                }
                else {
                    if (this.args.bizFlowchartId == null) {
                        list$ = this.dataService.getSlimCreationRulesBySourceVoIdAndTargetVoId(this.args.sourceVoId, this.args.targetVoId);
                    }
                    else {
                        list$ = this.dataService.getSlimCreationRulesBySourceVoIdAndTargetVoIdAndFlowchartId(this.args.sourceVoId, this.args.targetVoId, this.args.bizFlowchartId);
                    }
                }
                list$.subscribe(( /**
                 * @param {?} result
                 * @return {?}
                 */function (result) {
                    /** @type {?} */
                    var list = result;
                    list.forEach(( /**
                     * @param {?} elt
                     * @return {?}
                     */function (elt) {
                        if (_this.targetVoIdList.includes(elt.targetVoId) === false) {
                            _this.targetVoIdList.push(elt.targetVoId);
                            _this.creationRuleDict[elt.targetVoId] = new Array();
                        }
                        _this.targetVoNameDict[elt.targetVoId] = elt.targetVoName;
                        _this.creationRuleDict[elt.targetVoId].push(elt);
                    }));
                    _this.selectedTargetVoId = _this.targetVoIdList.length > 0 ? _this.targetVoIdList[0] : null;
                }));
            };
        PushSelectorComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'lib-push-selector',
                        template: "<div class=\"g-main p-2\">\r\n  <!--  -->\r\n  <div class=\"form-group\" style=\"height: 150px;overflow-y: auto;\">\r\n    <div [hidden]=\"targetVoIdList.length != 0\"><span style=\"color: red;\">{{ 'emptyResult' | aifLocale }}</span></div>\r\n    <div>{{ 'selectTargetBillType' | aifLocale }}</div>\r\n    <div *ngFor=\"let targetVoId of targetVoIdList;let i=index\" class=\"col-sm-12 col-md-5 form-check form-check-inline\">\r\n      <input class=\"form-check-input\" type=\"radio\" [name]=\"'targetVoId'+i\" id=\"targetVoId\" [value]=\"targetVoId\"\r\n        [(ngModel)]=\"selectedTargetVoId\" />\r\n      <label class=\"form-check-label\" [for]=\"'targetVoId'+i\">{{targetVoNameDict[targetVoId]}}</label>\r\n    </div>\r\n  </div>\r\n\r\n  <!--  -->\r\n  <div class=\"form-group\" style=\"\">\r\n    <div>{{ 'selectCreationRule' | aifLocale }}</div>\r\n    <select class=\"form-control\" name=\"selectedCreationRule\" [(ngModel)]=\"selectedCreationRule\">\r\n      <option *ngFor=\"let cr of creationRuleDict[selectedTargetVoId];\" [ngValue]=\"cr\">\r\n        {{cr.name}}</option>\r\n    </select>\r\n  </div>\r\n\r\n  <!--  -->\r\n  <div class=\"form-group\" style=\"\">\r\n    <div>{{ 'creationRuleDesciption' | aifLocale }}</div>\r\n    <textarea readonly class=\"form-control\" name=\"description\" rows=\"3\"\r\n      [ngModel]=\"selectedCreationRule&&selectedCreationRule.description\"></textarea>\r\n  </div>\r\n</div>",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        PushSelectorComponent.ctorParameters = function () {
            return [
                { type: undefined, decorators: [{ type: core.Inject, args: [SELECTOR_ARGS_INJECTOR,] }] },
                { type: AifDataService },
                { type: String, decorators: [{ type: core.Inject, args: [core.LOCALE_ID,] }] }
            ];
        };
        return PushSelectorComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @deprecated please use \@edp-bif/runtime-api
     */
    var SourceDataSelectorComponent = /** @class */ (function () {
        function SourceDataSelectorComponent(localeId) {
            var _this = this;
            this.subject4EntityDatas = new rxjs.Subject();
            this.cols = [];
            this.treedata = [];
            this.clickOk4AifSelector = ( /**
             * @return {?}
             */function () {
                _this.subject4EntityDatas.next([]);
                _this.subject4EntityDatas.complete();
                _this.bsModalRef4AifSelector.close();
            });
            this.clickClose4AifSelector = ( /**
             * @return {?}
             */function () {
                _this.bsModalRef4AifSelector.close();
            });
        }
        /**
         * @return {?}
         */
        SourceDataSelectorComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
            };
        SourceDataSelectorComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'lib-source-data-selector',
                        template: "<div class=\"g-main p-2\">\r\n  <farris-treetable #treetable [columns]=\"cols\" [showBorder]=\"false\" [showHeader]=\"false\" [data]=\"treedata\"\r\n    [idField]=\"'id'\" [fit]=\"true\" [singleSelect]=\"true\" [showCheckbox]=\"false\" [showIcon]=\"true\" [disabled]=\"false\">\r\n  </farris-treetable>\r\n\r\n  <ng-template #cell let-ctx>\r\n\r\n  </ng-template>\r\n\r\n  <p>\r\n    Hello world\r\n  </p>\r\n\r\n  <!-- <div class=\"form-group\" style=\"float: right;\">\r\n    <button class=\"btn-primary btn btn-sm ml-2\" (click)=\"clickOk4Aif()\">\u786E\u5B9A</button>\r\n    <button class=\"btn btn-sm ml-2 btn-default\" (click)=\"clickClose4Aif()\">\u53D6\u6D88</button>\r\n  </div> -->\r\n</div>",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        SourceDataSelectorComponent.ctorParameters = function () {
            return [
                { type: String, decorators: [{ type: core.Inject, args: [core.LOCALE_ID,] }] }
            ];
        };
        SourceDataSelectorComponent.propDecorators = {
            treetable: [{ type: core.ViewChild, args: ["treetable",] }]
        };
        return SourceDataSelectorComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @deprecated please use \@edp-bif/runtime-api
     * @type {?}
     */
    var LANG_RESOURCES = {
        "zh-CHS": {
            "confirm": "确定",
            "cancel": "取消",
            "emptyResult": "查询结果为空",
            "selectSourceBillType": "选择来源单据类型",
            "selectTargetBillType": "选择目标单据类型",
            "selectCreationRule": "选择生单规则",
            "creationRuleDesciption": "生单规则描述",
            "creationRuleSelector": "生单规则选择器",
            "sourceDataSelector": "来源单据选择器",
        },
        "en": {
            "confirm": "confirm",
            "cancel": "cancel",
            "emptyResult": "The result is empty",
            "selectSourceBillType": "Select source bill type",
            "selectTargetBillType": "Select target bill type",
            "selectCreationRule": "Select creation rule",
            "creationRuleDesciption": "Desciption",
            "creationRuleSelector": "CreationRule Selector",
            "sourceDataSelector": "SourceBill Selector",
        }
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @deprecated please use \@edp-bif/runtime-api
     */
    var AifLocalePipe = /** @class */ (function () {
        function AifLocalePipe(localeId) {
            this.langCode = localeId || "zh-CHS";
            this.lang = LANG_RESOURCES[this.langCode];
        }
        /**
         * @param {?} name
         * @return {?}
         */
        AifLocalePipe.prototype.transform = /**
         * @param {?} name
         * @return {?}
         */
            function (name) {
                return this.lang[name] || name;
            };
        AifLocalePipe.decorators = [
            { type: core.Pipe, args: [{ name: 'aifLocale' },] }
        ];
        /** @nocollapse */
        AifLocalePipe.ctorParameters = function () {
            return [
                { type: String, decorators: [{ type: core.Inject, args: [core.LOCALE_ID,] }] }
            ];
        };
        return AifLocalePipe;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ɵ0 = localStorage.getItem('languageCode') || 'zh-CHS', ɵ1 = {};
    /**
     * @deprecated please use \@edp-bif/runtime-api
     */
    var AifRuntimeModule = /** @class */ (function () {
        function AifRuntimeModule() {
        }
        /**
         * @param {?=} config
         * @return {?}
         */
        AifRuntimeModule.forRoot = /**
         * @param {?=} config
         * @return {?}
         */
            function (config) {
                return {
                    "ngModule": AifRuntimeModule,
                    "providers": [
                        { provide: AifRuntimeModuleConfig, useValue: config || {} },
                    ]
                };
            };
        AifRuntimeModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule,
                            forms.FormsModule,
                            forms.ReactiveFormsModule,
                            http.HttpClientModule,
                            uiModal.ModalModule.forRoot(),
                            uiNotify.NotifyModule.forRoot(),
                            uiMessager.MessagerModule.forRoot(),
                            uiDialog.FarrisDialogModule.forRoot(),
                            uiTreetable.TreeTableModule,
                        ],
                        providers: [
                            { provide: core.LOCALE_ID, useValue: ɵ0 },
                            { provide: AifRuntimeModuleConfig, useValue: ɵ1 },
                            { provide: core.NgModuleFactoryLoader, useClass: core.SystemJsNgModuleLoader },
                        ],
                        declarations: [
                            ReferSelectorComponent,
                            PushSelectorComponent,
                            SourceDataSelectorComponent,
                            AifLocalePipe,
                        ],
                        exports: [
                            ReferSelectorComponent,
                            PushSelectorComponent,
                            SourceDataSelectorComponent,
                        ],
                        entryComponents: [
                            ReferSelectorComponent,
                            PushSelectorComponent,
                            SourceDataSelectorComponent,
                        ]
                    },] }
        ];
        return AifRuntimeModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @deprecated please use \@edp-bif/runtime-api
     * @type {?}
     */
    var AIF_UISTATE_KEY_OPEN_TYPE = "abdbaec8-0bbe-4894-acee-b8d493bc4e1c";
    /**
     * @deprecated please use \@edp-bif/runtime-api
     * @type {?}
     */
    var AIF_UISTATE_KEY_CREATION_RULE_RESULT = "5eb32009-1d86-43df-96c8-8d685aea4e66";

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @abstract
     */
    var AbstractAifUiService = /** @class */ (function () {
        function AbstractAifUiService(frameContext, modalService, injector, cfr, loader, localeId) {
            this.frameContext = frameContext;
            this.modalService = modalService;
            this.injector = injector;
            this.cfr = cfr;
            this.loader = loader;
            this.localeId = localeId;
            /**
             * 内置模态框配置
             */
            this.defaultModalOptions = {
                title: null,
                width: 650,
                height: 415,
                buttons: null,
                showButtons: null,
            };
            this.aifLocalePipe = new AifLocalePipe(localeId);
        }
        /**
         * @protected
         * @param {?} args
         * @param {?} type
         * @param {?=} modalOptions
         * @param {?=} customSelectorModuleUrl
         * @return {?}
         */
        AbstractAifUiService.prototype.openCreationRuleSelector = /**
         * @protected
         * @param {?} args
         * @param {?} type
         * @param {?=} modalOptions
         * @param {?=} customSelectorModuleUrl
         * @return {?}
         */
            function (args, type, modalOptions, customSelectorModuleUrl) {
                var _this = this;
                if (typeof modalOptions == "string") {
                    if (modalOptions == "") {
                        throw "modalOptions不可为空";
                    }
                    modalOptions = JSON.parse(modalOptions);
                }
                modalOptions = modalOptions == null ? this.defaultModalOptions : modalOptions;
                modalOptions.title = modalOptions.title == null ? this.aifLocalePipe.transform("creationRuleSelector") : modalOptions.title;
                /** @type {?} */
                var cmpR = null;
                if (customSelectorModuleUrl == null || customSelectorModuleUrl.trim() == "") {
                    /** @type {?} */
                    var cmpF = this.cfr.resolveComponentFactory(type);
                    /** @type {?} */
                    var inj = core.ReflectiveInjector.resolveAndCreate([
                        { provide: SELECTOR_ARGS_INJECTOR, useValue: args.ConvertToJsonObject() },
                    ], this.injector);
                    cmpR = cmpF.create(inj);
                    this.configModalButtons(modalOptions, cmpR);
                    cmpR.instance.bsModalRef4AifSelector = this.modalService.show(cmpR, modalOptions);
                    return cmpR.instance.subject4CreationRule.asObservable();
                }
                else {
                    return rxjs.from(this.loader.load(customSelectorModuleUrl)).pipe(operators.switchMap(( /**
                     * @param {?} moduleFactory
                     * @return {?}
                     */function (moduleFactory) {
                        /** @type {?} */
                        var moduleRef = moduleFactory.create(_this.frameContext.root.injector);
                        cmpR = moduleRef.instance.createCreationRuleSelector(_this.frameContext.root.injector);
                        _this.configModalButtons(modalOptions, cmpR);
                        cmpR.instance.bsModalRef4AifSelector = _this.modalService.show(cmpR, modalOptions);
                        return cmpR.instance.subject4CreationRule.asObservable();
                    })));
                }
            };
        /**
         * 配置模态框按钮事件
         * @param modalOptions
         * @param cmpR
         */
        /**
         * 配置模态框按钮事件
         * @protected
         * @param {?} modalOptions
         * @param {?} cmpR
         * @return {?}
         */
        AbstractAifUiService.prototype.configModalButtons = /**
         * 配置模态框按钮事件
         * @protected
         * @param {?} modalOptions
         * @param {?} cmpR
         * @return {?}
         */
            function (modalOptions, cmpR) {
                modalOptions.buttons = [
                    {
                        text: this.aifLocalePipe.transform("confirm"),
                        cls: "btn btn-primary",
                        handle: cmpR.instance.clickOk4AifSelector,
                    },
                    {
                        text: this.aifLocalePipe.transform("cancel"),
                        cls: "btn btn-secondary",
                        handle: cmpR.instance.clickClose4AifSelector,
                    }
                ];
                modalOptions.showButtons = true;
            };
        /** @nocollapse */
        AbstractAifUiService.ctorParameters = function () {
            return [
                { type: devkit.FrameContext },
                { type: uiModal.BsModalService },
                { type: core.Injector },
                { type: core.ComponentFactoryResolver },
                { type: core.NgModuleFactoryLoader },
                { type: String, decorators: [{ type: core.Inject, args: [core.LOCALE_ID,] }] }
            ];
        };
        return AbstractAifUiService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var AifSourceUiService = /** @class */ (function (_super) {
        __extends(AifSourceUiService, _super);
        function AifSourceUiService(aifDataService, navigationService, frameContext, modalService, injector, cfr, loader, localeId) {
            var _this = _super.call(this, frameContext, modalService, injector, cfr, loader, localeId) || this;
            _this.aifDataService = aifDataService;
            _this.navigationService = navigationService;
            _this.frameContext = frameContext;
            _this.modalService = modalService;
            _this.injector = injector;
            _this.cfr = cfr;
            _this.loader = loader;
            _this.localeId = localeId;
            _this.aifUIStateCreationRuleResultKey = AIF_UISTATE_KEY_CREATION_RULE_RESULT;
            return _this;
        }
        /**
         * 下推时打开生单规则选择器
         */
        /**
         * 下推时打开生单规则选择器
         * @param {?} sourceVoId
         * @param {?=} bizFlowchartId
         * @param {?=} modalOptions
         * @param {?=} customSelectorModuleUrl
         * @return {?}
         */
        AifSourceUiService.prototype.openCreationRuleSelector4Push = /**
         * 下推时打开生单规则选择器
         * @param {?} sourceVoId
         * @param {?=} bizFlowchartId
         * @param {?=} modalOptions
         * @param {?=} customSelectorModuleUrl
         * @return {?}
         */
            function (sourceVoId, bizFlowchartId, modalOptions, customSelectorModuleUrl) {
                /** @type {?} */
                var args = new FormGetCreationRuleEntityExArgs();
                args.sourceVoId = sourceVoId;
                args.bizFlowchartId = bizFlowchartId;
                return this.openCreationRuleSelector(args, PushSelectorComponent, modalOptions, customSelectorModuleUrl);
            };
        /**
         * 特定下推时打开生单规则选择器
         */
        /**
         * 特定下推时打开生单规则选择器
         * @param {?} sourceVoId
         * @param {?} targetVoId
         * @param {?=} bizFlowchartId
         * @param {?=} modalOptions
         * @param {?=} customSelectorModuleUrl
         * @return {?}
         */
        AifSourceUiService.prototype.openCreationRuleSelector4SpecificPush = /**
         * 特定下推时打开生单规则选择器
         * @param {?} sourceVoId
         * @param {?} targetVoId
         * @param {?=} bizFlowchartId
         * @param {?=} modalOptions
         * @param {?=} customSelectorModuleUrl
         * @return {?}
         */
            function (sourceVoId, targetVoId, bizFlowchartId, modalOptions, customSelectorModuleUrl) {
                /** @type {?} */
                var args = new FormGetCreationRuleEntityExArgs();
                args.sourceVoId = sourceVoId;
                args.targetVoId = targetVoId;
                args.bizFlowchartId = bizFlowchartId;
                return this.openCreationRuleSelector(args, PushSelectorComponent, modalOptions, customSelectorModuleUrl);
            };
        /**
         * 打开下游单据制单界面
         * @param formSettings
         * @param rpcCreationRuleExecutorResult
         * @param extParams
         */
        /**
         * 打开下游单据制单界面
         * @param {?} formSettings
         * @param {?} rpcCreationRuleExecutorResult
         * @param {?=} extParams
         * @return {?}
         */
        AifSourceUiService.prototype.openTargetForm = /**
         * 打开下游单据制单界面
         * @param {?} formSettings
         * @param {?} rpcCreationRuleExecutorResult
         * @param {?=} extParams
         * @return {?}
         */
            function (formSettings, rpcCreationRuleExecutorResult, extParams) {
                var e_1, _a;
                if (formSettings == null) {
                    throw "参数formSettings不可为空";
                }
                if (rpcCreationRuleExecutorResult == null) {
                    throw "参数rpcCreationRuleExecutorResult不可为空";
                }
                if (typeof rpcCreationRuleExecutorResult == "string") {
                    if (rpcCreationRuleExecutorResult == "") {
                        throw "参数rpcCreationRuleExecutorResult不可为空";
                    }
                    /** @type {?} */
                    var temp = new RpcCreationRuleExecutorResult();
                    temp.LoadFromJson(rpcCreationRuleExecutorResult);
                    rpcCreationRuleExecutorResult = temp;
                }
                if ((rpcCreationRuleExecutorResult instanceof RpcCreationRuleExecutorResult) == false) {
                    /** @type {?} */
                    var res = new RpcCreationRuleExecutorResult();
                    res.LoadFromJsonObject(rpcCreationRuleExecutorResult);
                    rpcCreationRuleExecutorResult = res;
                }
                if (rpcCreationRuleExecutorResult.mappingResults == null || rpcCreationRuleExecutorResult.mappingResults.length == 0) {
                    return;
                }
                else if (rpcCreationRuleExecutorResult.mappingResults.length == 1) {
                    //只有一条数据则打开卡片
                    this.navigationService.openApp(rpcCreationRuleExecutorResult.mappingResults[0].associations[0].tarDataId, formSettings.targetCardAppId, formSettings.targetCardAppEntrance, this.buildRouteParams(rpcCreationRuleExecutorResult, extParams == null ? {} : extParams), false);
                }
                else {
                    if (formSettings.targetListAppId != null) {
                        // 多条数据且配置批量编辑界面
                        throw "暂不支持打开批量编辑界面";
                        this.navigationService.openApp(rpcCreationRuleExecutorResult.mappingResults[0].associations[0].tarDataId, formSettings.targetListAppId, formSettings.targetListAppEntrance, this.buildRouteParams(rpcCreationRuleExecutorResult, extParams == null ? {} : extParams), false);
                    }
                    else {
                        try {
                            //有多条数据但未配置批量编辑界面则打开多个卡片
                            for (var _b = __values(rpcCreationRuleExecutorResult.mappingResults), _c = _b.next(); !_c.done; _c = _b.next()) {
                                var mappingResult = _c.value;
                                /** @type {?} */
                                var singleRes = new RpcCreationRuleExecutorResult();
                                singleRes.bizFlowchartId = rpcCreationRuleExecutorResult.bizFlowchartId;
                                singleRes.creationRuleId = rpcCreationRuleExecutorResult.creationRuleId;
                                singleRes.creationRuleInstanceId = rpcCreationRuleExecutorResult.creationRuleInstanceId;
                                singleRes.mappingResults = [mappingResult];
                                this.navigationService.openApp(mappingResult.associations[0].tarDataId, formSettings.targetCardAppId, formSettings.targetCardAppEntrance, this.buildRouteParams(singleRes, extParams == null ? {} : extParams), false);
                            }
                        }
                        catch (e_1_1) {
                            e_1 = { error: e_1_1 };
                        }
                        finally {
                            try {
                                if (_c && !_c.done && (_a = _b.return))
                                    _a.call(_b);
                            }
                            finally {
                                if (e_1)
                                    throw e_1.error;
                            }
                        }
                    }
                }
            };
        /**
         *
         * @param action 目标App动作
         * @param pairs 其他实体参数键值对
         */
        /**
         *
         * @private
         * @param {?} rpcCreationRuleExecutorResult
         * @param {?} extParams
         * @return {?}
         */
        AifSourceUiService.prototype.buildRouteParams = /**
         *
         * @private
         * @param {?} rpcCreationRuleExecutorResult
         * @param {?} extParams
         * @return {?}
         */
            function (rpcCreationRuleExecutorResult, extParams) {
                /** @type {?} */
                var params = extParams;
                params[this.aifUIStateCreationRuleResultKey] = rpcCreationRuleExecutorResult;
                return params;
            };
        AifSourceUiService.decorators = [
            { type: core.Injectable }
        ];
        /** @nocollapse */
        AifSourceUiService.ctorParameters = function () {
            return [
                { type: AifDataService },
                { type: commandServices.NavigationService },
                { type: devkit.FrameContext },
                { type: uiModal.BsModalService },
                { type: core.Injector },
                { type: core.ComponentFactoryResolver },
                { type: core.NgModuleFactoryLoader },
                { type: String, decorators: [{ type: core.Inject, args: [core.LOCALE_ID,] }] }
            ];
        };
        return AifSourceUiService;
    }(AbstractAifUiService));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var AifTargetUiService = /** @class */ (function (_super) {
        __extends(AifTargetUiService, _super);
        function AifTargetUiService(aifDataService, frameContext, modalService, injector, cfr, loader, localeId) {
            var _this = _super.call(this, frameContext, modalService, injector, cfr, loader, localeId) || this;
            _this.aifDataService = aifDataService;
            _this.frameContext = frameContext;
            _this.modalService = modalService;
            _this.injector = injector;
            _this.cfr = cfr;
            _this.loader = loader;
            _this.localeId = localeId;
            _this.aifUIStateCreationRuleResultKey = AIF_UISTATE_KEY_CREATION_RULE_RESULT;
            return _this;
        }
        /**
         * 参照时打开生单规则选择器
         */
        /**
         * 参照时打开生单规则选择器
         * @param {?} targetVoId
         * @param {?=} bizFlowchartId
         * @param {?=} modalOptions
         * @param {?=} customSelectorModuleUrl
         * @return {?}
         */
        AifTargetUiService.prototype.openCreationRuleSelector4Refer = /**
         * 参照时打开生单规则选择器
         * @param {?} targetVoId
         * @param {?=} bizFlowchartId
         * @param {?=} modalOptions
         * @param {?=} customSelectorModuleUrl
         * @return {?}
         */
            function (targetVoId, bizFlowchartId, modalOptions, customSelectorModuleUrl) {
                /** @type {?} */
                var args = new FormGetCreationRuleEntityExArgs();
                args.targetVoId = targetVoId;
                args.bizFlowchartId = bizFlowchartId;
                return this.openCreationRuleSelector(args, ReferSelectorComponent, modalOptions, customSelectorModuleUrl);
            };
        /**
         * 特定参照时打开生单规则选择器
         */
        /**
         * 特定参照时打开生单规则选择器
         * @param {?} sourceVoId
         * @param {?} targetVoId
         * @param {?=} bizFlowchartId
         * @param {?=} modalOptions
         * @param {?=} customSelectorModuleUrl
         * @return {?}
         */
        AifTargetUiService.prototype.openCreationRuleSelector4SpecificRefer = /**
         * 特定参照时打开生单规则选择器
         * @param {?} sourceVoId
         * @param {?} targetVoId
         * @param {?=} bizFlowchartId
         * @param {?=} modalOptions
         * @param {?=} customSelectorModuleUrl
         * @return {?}
         */
            function (sourceVoId, targetVoId, bizFlowchartId, modalOptions, customSelectorModuleUrl) {
                /** @type {?} */
                var args = new FormGetCreationRuleEntityExArgs();
                args.sourceVoId = sourceVoId;
                args.targetVoId = targetVoId;
                args.bizFlowchartId = bizFlowchartId;
                return this.openCreationRuleSelector(args, ReferSelectorComponent, modalOptions, customSelectorModuleUrl);
            };
        /**
         * 打开来源单据选择器
         * @param formSettings
         * @param modalOptions
         */
        /**
         * 打开来源单据选择器
         * @param {?=} formSettings
         * @param {?=} modalOptions
         * @return {?}
         */
        AifTargetUiService.prototype.openSourceDataSelector4Refer = /**
         * 打开来源单据选择器
         * @param {?=} formSettings
         * @param {?=} modalOptions
         * @return {?}
         */
            function (formSettings, modalOptions) {
                var _this = this;
                modalOptions = modalOptions == null ? this.defaultModalOptions : modalOptions;
                modalOptions.title = modalOptions.title == null ? this.aifLocalePipe.transform("sourceDataSelector") : modalOptions.title;
                /** @type {?} */
                var cmpR = null;
                if (formSettings == null || formSettings.sourceDataSelectorModuleUrl == null) {
                    throw "暂无默认实现，请使用自定义来源单据选择器";
                    /** @type {?} */
                    var cmpF = this.cfr.resolveComponentFactory(SourceDataSelectorComponent);
                    /** @type {?} */
                    var inj = core.Injector.create([], this.injector);
                    cmpR = cmpF.create(inj);
                    this.configModalButtons(modalOptions, cmpR);
                    cmpR.instance.bsModalRef4AifSelector = this.modalService.show(cmpR, modalOptions);
                    return cmpR.instance.subject4EntityDatas.asObservable();
                }
                else {
                    // return from(System.import(formSettings.sourceDataSelectorModuleUrl.substring(0, formSettings.sourceDataSelectorModuleUrl.indexOf("#")) + ".js")).pipe(
                    return rxjs.from(this.loader.load(formSettings.sourceDataSelectorModuleUrl)).pipe(operators.switchMap(( /**
                     * @param {?} moduleFactory
                     * @return {?}
                     */function (moduleFactory) {
                        /** @type {?} */
                        var moduleRef = moduleFactory.create(_this.frameContext.root.injector);
                        cmpR = moduleRef.instance.createSourceDataSelector(_this.frameContext.root.injector);
                        _this.configModalButtons(modalOptions, cmpR);
                        cmpR.instance.bsModalRef4AifSelector = _this.modalService.show(cmpR, modalOptions);
                        return cmpR.instance.subject4EntityDatas.asObservable();
                    })));
                }
            };
        /**
         * 获取传到根组件UIState上的生单规则执行器结果
         */
        /**
         * 获取传到根组件UIState上的生单规则执行器结果
         * @return {?}
         */
        AifTargetUiService.prototype.getRpcCreationRuleExecutorResult = /**
         * 获取传到根组件UIState上的生单规则执行器结果
         * @return {?}
         */
            function () {
                /** @type {?} */
                var temp = this.frameContext.root.uiState[this.aifUIStateCreationRuleResultKey];
                /** @type {?} */
                var result = new RpcCreationRuleExecutorResult();
                if (typeof temp == "string") {
                    result.LoadFromJson(temp);
                }
                else if (typeof temp == "object") {
                    result.LoadFromJsonObject(temp);
                }
                else {
                    throw "不合法的类型";
                }
                return result;
            };
        AifTargetUiService.decorators = [
            { type: core.Injectable }
        ];
        /** @nocollapse */
        AifTargetUiService.ctorParameters = function () {
            return [
                { type: AifDataService },
                { type: devkit.FrameContext },
                { type: uiModal.BsModalService },
                { type: core.Injector },
                { type: core.ComponentFactoryResolver },
                { type: core.NgModuleFactoryLoader },
                { type: String, decorators: [{ type: core.Inject, args: [core.LOCALE_ID,] }] }
            ];
        };
        return AifTargetUiService;
    }(AbstractAifUiService));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.AifManualCreationType = AifManualCreationType;
    exports.AifRuntimeModuleConfig = AifRuntimeModuleConfig;
    exports.AifRuntimeModule = AifRuntimeModule;
    exports.AifDataService = AifDataService;
    exports.AifSourceUiService = AifSourceUiService;
    exports.AifTargetUiService = AifTargetUiService;
    exports.SourceDataSelectorComponent = SourceDataSelectorComponent;
    exports.PushSelectorComponent = PushSelectorComponent;
    exports.ReferSelectorComponent = ReferSelectorComponent;
    exports.BefContentVariable = BefContentVariable;
    exports.CreationMappingResult = CreationMappingResult;
    exports.CreationRuleExecutorArgs = CreationRuleExecutorArgs;
    exports.CreationRuleExecutorAssociations = CreationRuleExecutorAssociations;
    exports.CreationRuleExecutorResult = CreationRuleExecutorResult;
    exports.MappingAssociation = MappingAssociation;
    exports.MappingResult = MappingResult;
    exports.ConditionResult = ConditionResult;
    exports.FormBillCreationProcessExecutorResult = FormBillCreationProcessExecutorResult;
    exports.RpcCreationMappingResult = RpcCreationMappingResult;
    exports.RpcCreationRuleExecutorArgs = RpcCreationRuleExecutorArgs;
    exports.RpcCreationRuleExecutorResult = RpcCreationRuleExecutorResult;
    exports.RpcFireBizEventArgs = RpcFireBizEventArgs;
    exports.RpcMappingResult = RpcMappingResult;
    exports.RpcVoChangeDetail = RpcVoChangeDetail;
    exports.RpcWritebackMappingResult = RpcWritebackMappingResult;
    exports.RpcWritebackRuleExecutorArgs = RpcWritebackRuleExecutorArgs;
    exports.RpcWritebackRuleExecutorResult = RpcWritebackRuleExecutorResult;
    exports.FormCreateArgs = FormCreateArgs;
    exports.FormCreateResult = FormCreateResult;
    exports.FormCreationRuleExecutorArgs = FormCreationRuleExecutorArgs;
    exports.FormEntityData = FormEntityData;
    exports.FormRestArgs = FormRestArgs;
    exports.FormRestResult = FormRestResult;
    exports.FormUpdateArgs = FormUpdateArgs;
    exports.FormUpdateResult = FormUpdateResult;
    exports.FormGetCreationRuleEntityExArgs = FormGetCreationRuleEntityExArgs;
    exports.FormGetCreationRuleEntityExResult = FormGetCreationRuleEntityExResult;
    exports.AIF_UISTATE_KEY_OPEN_TYPE = AIF_UISTATE_KEY_OPEN_TYPE;
    exports.AIF_UISTATE_KEY_CREATION_RULE_RESULT = AIF_UISTATE_KEY_CREATION_RULE_RESULT;
    exports.ɵa = SELECTOR_ARGS_INJECTOR;
    exports.ɵb = AifLocalePipe;
    exports.ɵc = AbstractAifUiService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
