// @gsp-wf/wf-process-editor/esm2015/lib/views/aif-ui-common/pipe/aif-ui-common-locale.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe, Inject, LOCALE_ID, InjectionToken } from '@angular/core';
/** @type {?} */
export const BIF_LANG_RESOURCES = new InjectionToken("BIF_LANG_RESOURCES");
export class AifLocalePipe {
    /**
     * @param {?} localeId
     * @param {?} resources
     */
    constructor(localeId, resources) {
        this.defaultLang = "zh-CHS";
        this.langCode = localeId || this.defaultLang;
        this.resources = resources || {};
        this.lang = resources[this.langCode] || resources[this.defaultLang] || {};
    }
    /**
     * @param {?} name
     * @return {?}
     */
    transform(name) {
        return this.lang[name] || name;
    }
    /**
     * @param {?} name
     * @return {?}
     */
    instant(name) {
        return this.transform(name);
    }
}
AifLocalePipe.decorators = [
    { type: Pipe, args: [{ name: 'aifLocale' },] }
];
/** @nocollapse */
AifLocalePipe.ctorParameters = () => [
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [BIF_LANG_RESOURCES,] }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AifLocalePipe.prototype.defaultLang;
    /**
     * @type {?}
     * @private
     */
    AifLocalePipe.prototype.langCode;
    /**
     * @type {?}
     * @private
     */
    AifLocalePipe.prototype.lang;
    /**
     * @type {?}
     * @private
     */
    AifLocalePipe.prototype.resources;
}
