// @farris/ui-datagrid/esm2015/lib/scrollbar/scrollbar.interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/** @type {?} */
export const SCROLLBAR_CONFIG = new InjectionToken('PERFECT_SCROLLBAR_CONFIG');
export class Geometry {
    /**
     * @param {?} x
     * @param {?} y
     * @param {?} w
     * @param {?} h
     */
    constructor(x, y, w, h) {
        this.x = x;
        this.y = y;
        this.w = w;
        this.h = h;
    }
}
if (false) {
    /** @type {?} */
    Geometry.prototype.x;
    /** @type {?} */
    Geometry.prototype.y;
    /** @type {?} */
    Geometry.prototype.w;
    /** @type {?} */
    Geometry.prototype.h;
}
export class Position {
    /**
     * @param {?} x
     * @param {?} y
     */
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
}
if (false) {
    /** @type {?} */
    Position.prototype.x;
    /** @type {?} */
    Position.prototype.y;
}
/** @type {?} */
export const ScrollbarEvents = [
    'psScrollY',
    'psScrollX',
    'psScrollUp',
    'psScrollDown',
    'psScrollLeft',
    'psScrollRight',
    'psYReachEnd',
    'psYReachStart',
    'psXReachEnd',
    'psXReachStart'
];
/**
 * @record
 */
export function ScrollbarConfigInterface() { }
if (false) {
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.handlers;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.wheelSpeed;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.swipeEasing;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.suppressScrollX;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.suppressScrollY;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.wheelPropagation;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.useBothWheelAxes;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.scrollingThreshold;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.minScrollbarLength;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.maxScrollbarLength;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.scrollXMarginOffset;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.scrollYMarginOffset;
}
export class ScrollbarConfig {
    /**
     * @param {?=} config
     */
    constructor(config = {}) {
        this.assign(config);
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    assign(config = {}) {
        for (const key in config) {
            this[(/** @type {?} */ (key))] = config[(/** @type {?} */ (key))];
        }
    }
}
if (false) {
    /** @type {?} */
    ScrollbarConfig.prototype.handlers;
    /** @type {?} */
    ScrollbarConfig.prototype.wheelSpeed;
    /** @type {?} */
    ScrollbarConfig.prototype.swipeEasing;
    /** @type {?} */
    ScrollbarConfig.prototype.suppressScrollX;
    /** @type {?} */
    ScrollbarConfig.prototype.suppressScrollY;
    /** @type {?} */
    ScrollbarConfig.prototype.wheelPropagation;
    /** @type {?} */
    ScrollbarConfig.prototype.useBothWheelAxes;
    /** @type {?} */
    ScrollbarConfig.prototype.scrollingThreshold;
    /** @type {?} */
    ScrollbarConfig.prototype.minScrollbarLength;
    /** @type {?} */
    ScrollbarConfig.prototype.maxScrollbarLength;
    /** @type {?} */
    ScrollbarConfig.prototype.scrollXMarginOffset;
    /** @type {?} */
    ScrollbarConfig.prototype.scrollYMarginOffset;
}
