// @farris/ide-publish-menu/esm5/lib/locale.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var PUBLISH_MENU_LOCALE = {
    "zh-CHS": {
        publishmenu: {
            app: {
                id: '应用标识',
                code: '应用编号',
                name: '应用名称',
                path: '部署路径',
                remark: '应用描述',
                entry: '应用入口',
                entryname: '入口名称',
                entrycode: '入口编号',
                routeid: '路由标识',
                action: '操作',
                addentry: '新增入口',
                message: {
                    info: '选择上级应用',
                    appInvoks: '请设置应用入口信息',
                    inValidInvoks: '应用入口点信息不完整，请检查',
                    appcode: '应用编号不能为空!<br>',
                    appname: '应用名称不能为空!<br>',
                    url: '部署路径不能为空!<br>',
                    invoksCode: '入口名称不能为空!<br>',
                    invoksName: '入口编号不能为空!<br>',
                    invoksAppEntrance: '路由标识不能为空!<br>'
                },
                column: {
                    code: '编号',
                    name: '名称'
                }
            },
            func: {
                id: '功能标识',
                code: '功能编号',
                name: '功能名称',
                action: '功能操作',
                selectBiz: '选择',
                remark: '功能描述',
                selectapp: '选择',
                addapp: '新建',
                linkapp: '关联应用',
                appentry: '应用入口',
                appParam: '参数设置',
                paramName: '参数名称',
                paramValue: '参数值',
                paramAction: '操作',
                addParam: '新增参数',
                placeholder: {
                    code: '请输入功能编号',
                    name: '请输入功能名称',
                    biz: '选择业务对象操作',
                    remark: '请输入功能描述',
                    selectapp: '选择应用'
                },
                message: {
                    info: '选择上级菜单',
                    code: '功能编号不能为空！<br>',
                    name: '功能名称不能为空！<br>',
                    parentId: '请在左侧选择上级菜单！<br>',
                    appid: '请选择关联应用！<br>',
                    appInvokId: '请选择关联应用入口！<br>',
                    bizObjectId: '请选择左侧业务操作对象！ <br>',
                    bizOpId: '请选择功能操作！ <br>'
                }
            },
            dialog: {
                title: '发布新菜单',
                btnOk: '确定',
                btnCancel: '取消',
                success: '新菜单发布成功！',
                selectapp: '选择关联应用',
                selectInfo: '请选择应用',
                appentry: '请选择应用入口',
                bizOptitle: '选择业务操作',
                bizwarn: '请选择左侧业务操作对象',
                bizaction: '请选择业务操作',
                app: '应用管理'
            },
            line: '第 {line} 行'
        }
    },
    "zh-CHT": {
        publishmenu: {
            app: {
                id: '應用標識',
                code: '應用編號',
                name: '應用名稱',
                path: '部署路徑',
                entry: '應用入口',
                entryname: '入口名稱',
                entrycode: '入口編號',
                routeid: '路由標識',
                remark: '應用描述',
                action: '操作',
                addentry: '新增入口',
                message: {
                    info: '選擇上級應用',
                    appInvoks: '請設置應用入口信息',
                    inValidInvoks: '應用入口信息不完整，請檢查',
                    appcode: '應用編號不能為空！<br>',
                    appname: '應用名稱不能為空！<br>',
                    url: '部署路徑不能為空！<br>',
                    invoksCode: '入口名稱不能為空！<br>',
                    invoksName: '入口編號不能為空！<br>',
                    invoksAppEntrance: '路由標識不能為空！<br>'
                },
                column: {
                    code: '編號',
                    name: '名稱'
                }
            },
            func: {
                id: '功能標識',
                code: '功能編號',
                name: '功能名稱',
                action: '功能操作',
                selectBiz: '選擇',
                remark: '功能描述',
                selectapp: '選擇',
                addapp: '新建',
                linkapp: '關聯應用',
                appentry: '應用入口',
                appParam: '參數設置',
                paramName: '參數名稱',
                paramValue: '參數值',
                paramAction: '操作',
                addParam: '新增參數',
                placeholder: {
                    code: '請輸入功能編號',
                    name: '請輸入功能名稱',
                    biz: '選擇業務對象操作',
                    remark: '請輸入功能描述',
                    selectapp: '選擇應用'
                },
                message: {
                    info: '選擇上級菜單',
                    code: '功能編號不能為空！<br>',
                    name: '功能名稱不能為空！<br>',
                    parentId: '請在左側選擇上級菜單！<br>',
                    appid: '請選擇關聯應用！<br>',
                    appInvokId: '請選擇關聯應用入口！<br>',
                    bizObjectId: '請選擇左側業務操作對象！ <br>',
                    bizOpId: '請選擇功能操作！ <br>'
                }
            },
            dialog: {
                title: '發布新菜單',
                btnOk: '確定',
                btnCancel: '取消',
                success: '新菜單發布成功！',
                selectapp: '選擇關聯應用',
                selectInfo: '請選擇應用',
                appentry: '請選擇應用入口',
                bizOptitle: '選擇業務操作',
                bizwarn: '請選擇左側業務操作對象',
                bizaction: '請選擇業務操作',
                app: '應用管理'
            },
            line: '第 {line} 行'
        }
    },
    en: {
        publishmenu: {
            app: {
                id: 'AppId',
                code: 'AppCode',
                name: 'AppName',
                path: 'DeploymentPayh',
                entry: 'ProtalSettings',
                entryname: 'Protal Name',
                entrycode: 'Protal Code',
                routeid: 'RoutingID',
                action: 'Operation',
                remark: 'AppDesc',
                addentry: 'New Protal',
                message: {
                    info: 'Please select parent.<br>',
                    appInvoks: 'Please set app protal.<br>',
                    inValidInvoks: 'Please check the protal info.<br>',
                    appcode: 'The App Code should not be empty.<br>',
                    appname: 'The App Name should not be empty.<br>',
                    url: 'The Url should not be empty.<br>',
                    invoksCode: 'Entry name cannot be empty.<br>',
                    invoksName: 'Entrance number cannot be empty.<br>',
                    invoksAppEntrance: 'Route ID cannot be empty.<br>'
                },
                column: {
                    code: 'Code',
                    name: 'Name'
                }
            },
            func: {
                id: 'Function ID',
                code: 'Function code',
                name: 'Function name',
                action: 'Function operation',
                selectBiz: 'Select',
                remark: 'Function description',
                selectapp: 'Select App',
                addapp: 'New App',
                linkapp: 'Invoke information',
                appentry: 'ProtalSettings',
                appParam: 'Static params setting',
                paramName: 'Param Name',
                paramValue: 'Param Value',
                paramAction: '	Operation',
                addParam: 'Add new params',
                placeholder: {
                    code: 'Function code',
                    name: 'Function Name',
                    biz: 'Please Select Business Type.',
                    remark: 'Function description',
                    selectapp: 'Select App'
                },
                message: {
                    info: 'Please aelect a parent menu.',
                    code: 'The function Code should not be empty！<br>',
                    name: 'The function Name should not be empty！<br>',
                    parentId: 'Please select the upper menu on the left！<br>',
                    appid: 'Please select the associated application！<br>',
                    appInvokId: 'Please select the associated application portal！<br>',
                    bizObjectId: 'Please select the business operation object on the left！ <br>',
                    bizOpId: 'Please select function operation！ <br>'
                }
            },
            dialog: {
                title: 'New menu publish',
                btnOk: 'OK',
                btnCancel: 'Cancel',
                success: 'New menu publish success！',
                selectapp: 'Select associated application',
                selectInfo: 'Please select a application.',
                appentry: 'Please select an application portal.',
                bizOptitle: 'Select business operation.',
                bizwarn: 'Please select the business operation object on the left.',
                bizaction: 'Please select business operation.',
                app: 'Application Manager'
            },
            line: 'Line {line} '
        }
    }
};
