// @edp-pmf/bpmn-model/dist/cjs/lib/model/omgdc/Bounds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bounds = void 0;
var tslib_1 = require("tslib");
var BpmnModelHelper_1 = require("../BpmnModelHelper");
var OmgdcElement_1 = require("./OmgdcElement");
var Bounds = /** @class */ (function (_super) {
    tslib_1.__extends(Bounds, _super);
    function Bounds(model) {
        return _super.call(this, model) || this;
    }
    Bounds.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Height = 0;
        this.Width = 0;
        this.X = 0;
        this.Y = 0;
    };
    Bounds.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Height", this.Height);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Width", this.Width);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "X", this.X);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Y", this.Y);
        return obj;
    };
    Bounds.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Height = obj.Height;
        this.Width = obj.Width;
        this.X = obj.X;
        this.Y = obj.Y;
    };
    return Bounds;
}(OmgdcElement_1.OmgdcElement));
exports.Bounds = Bounds;
