// @edp-pmf/bpmn-model/dist/cjs/lib/model/omgdc/Point.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Point = void 0;
var tslib_1 = require("tslib");
var BpmnModelHelper_1 = require("../BpmnModelHelper");
var OmgdcElement_1 = require("./OmgdcElement");
var Point = /** @class */ (function (_super) {
    tslib_1.__extends(Point, _super);
    function Point(model) {
        return _super.call(this, model) || this;
    }
    Point.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.X = 0;
        this.Y = 0;
    };
    Point.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "X", this.X);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Y", this.Y);
        return obj;
    };
    Point.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.X = obj.X;
        this.Y = obj.Y;
    };
    return Point;
}(OmgdcElement_1.OmgdcElement));
exports.Point = Point;
