// @farris/devkit/esm2015/lib/app/binding_data_manager.js
/*
 * @Author: Witt
 * @Date: 2018-12-29 10:46:01
 * @Last Modified by: Witt
 * @Last Modified time: 2018-12-30 17:56:02
 */
/**
 * BindingData管理类
 */
class BindingDataManager {
    /**
     * 构造函数
     */
    constructor() {
        this.bindingDataMap = new Map();
    }
    /**
     * 获取BindingDataMap
     */
    getBindingDataMap() {
        return this.bindingDataMap;
    }
    /**
     * 根据name获取BindingData
     * @return 找不到时返回undefined
     */
    getBindingDataByName(name) {
        return this.bindingDataMap.get(name);
    }
    /**
     * 初始化全局的BindingData
     * @description 仅在高速模式时注册bindingData
     */
    regBindingData(name, bindingData) {
        this.bindingDataMap.set(name, bindingData);
    }
    unRegisteBindingData(name) {
        this.bindingDataMap.delete(name);
    }
    /**
     * 是否Repository已经存在
     */
    ifBindingDataExits(name) {
        const bindingData = this.getBindingDataByName(name);
        return bindingData ? true : false;
    }
    dispose() {
        this.bindingDataMap.clear();
    }
}
export { BindingDataManager };
