// @farris/devkit/esm2015/lib/app/repository_mananger.js
/*
 * @Author: Witt
 * @Date: 2018-12-29 10:46:01
 * @Last Modified by: Witt
 * @Last Modified time: 2018-12-30 18:06:11
 */
import { Injectable } from '@angular/core';
/**
 * Repository管理类
 */
class RepositoryManager {
    /**
     * 构造函数
     */
    constructor() {
        this.repositoryMap = new Map();
    }
    /**
     * 注册Repository
     */
    regRepository(name, repository) {
        this.repositoryMap.set(name, repository);
    }
    unRegisteRepository(name) {
        this.repositoryMap.delete(name);
    }
    /**
     * 获取RepositoryMap
     * @internal
     */
    getRepositoryMap() {
        return this.repositoryMap;
    }
    /**
     * 获取Repository数组
     */
    getRepositories() {
        return Array.from(this.repositoryMap.values());
    }
    /**
     * 根据name获取Repository
     */
    getRepositoryByName(name) {
        return this.repositoryMap.get(name);
    }
    /**
     * 是否Repository已经存在
     */
    ifRepositoryExits(name) {
        const repository = this.getRepositoryByName(name);
        return repository ? true : false;
    }
    dispose() {
        // this.repositoryMap.forEach((repo: Repository<any>) => {
        //   if (repo) {
        //     repo.entityCollection.reset(false);
        //   }
        // });
        this.repositoryMap.clear();
    }
}
RepositoryManager.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RepositoryManager.ctorParameters = () => [];
export { RepositoryManager };
