// @farris/ui-filter-editor/esm5/lib/services/filter.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { Subject } from 'rxjs';
import { FilterRelation } from '@farris/ui-common/types';
import { CommonUtils } from '@farris/ui-common';
var FilterService = /** @class */ (function () {
    function FilterService(utils) {
        this.conditionsChanged = new Subject();
        this.commonUtils = utils || new CommonUtils();
    }
    /**
     * @param {?} c
     * @return {?}
     */
    FilterService.prototype.buildSqlWhere = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.commonUtils.buildSqlWhere(c);
    };
    /**
     * @param {?} conditions
     * @param {?=} localeService
     * @return {?}
     */
    FilterService.prototype.isValid = /**
     * @param {?} conditions
     * @param {?=} localeService
     * @return {?}
     */
    function (conditions, localeService) {
        if (conditions && conditions.length) {
            // 清除最后条件的关系符
            conditions[conditions.length - 1].relation = FilterRelation.Empty;
            // 检查关系连接符
            /** @type {?} */
            var emptyRelations = conditions
                .map((/**
             * @param {?} n
             * @param {?} i
             * @return {?}
             */
            function (n, i) {
                if (n.relation == "" + FilterRelation.Empty) {
                    return { index: i, filter: n };
                }
                return null;
            }))
                .filter((/**
             * @param {?} n
             * @param {?} i
             * @return {?}
             */
            function (n, i) {
                return n && i !== conditions.length - 1;
            }));
            /** @type {?} */
            var relationFlag = true;
            if (emptyRelations.length) {
                relationFlag = false;
            }
            // 检查括号
            /** @type {?} */
            var bracketFlag = true;
            /** @type {?} */
            var bracketStr = '';
            for (var i = 0; i < conditions.length; i++) {
                bracketStr += conditions[i].lbracket + conditions[i].rbracket;
            }
            /** @type {?} */
            var l = bracketStr.length;
            if (l % 2 > 0) {
                bracketFlag = false;
            }
            else {
                for (var i = 0; i < l; i++) {
                    bracketStr = bracketStr.replace('()', '');
                }
                bracketFlag = !bracketStr.length;
            }
            // 检查字段是否设置
            /** @type {?} */
            var emptyFields = conditions.filter((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return !n.filterField; })).length === 0;
            return {
                bracket: {
                    result: bracketFlag,
                    msg: localeService ? localeService.getValue('filterEditor.validate.bracket') : '括号不匹配，请检查。'
                },
                relation: {
                    result: relationFlag,
                    msg: localeService ? localeService.getValue('filterEditor.validate.relation') : '关系设置不完整，请检查。'
                },
                field: {
                    result: emptyFields,
                    msg: localeService ? localeService.getValue('filterEditor.validate.field') : '字段不允许为空，请检查。'
                }
            };
        }
        return true;
    };
    FilterService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FilterService.ctorParameters = function () { return [
        { type: CommonUtils, decorators: [{ type: Optional }] }
    ]; };
    return FilterService;
}());
export { FilterService };
if (false) {
    /** @type {?} */
    FilterService.prototype.conditionsChanged;
    /** @type {?} */
    FilterService.prototype.commonUtils;
}
