// @gspwidget/widget-devkit/esm2015/lib/preset/preset.echarts.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PlThemeColors } from "@gspwidget/util";
import { PropertyTypes } from "../../types";
import LinearGradient from "./zrender.linear-gradient";
/** @type {?} */
export const EchartsOptions = {
    grid: {
        top: 20,
        right: 30,
        bottom: 36,
        left: 10,
        containLabel: true,
    },
    xAxis: {
        type: 'category',
        axisLine: {
            lineStyle: {
                color: '#D9D9D9',
            }
        },
        axisTick: {
            lineStyle: {
                color: '#D9D9D9',
            }
        },
        axisLabel: {
            color: 'black',
            margin: 10,
            fontSize: 11
        }
    },
    yAxis: {
        axisLine: {
            show: false
        },
        axisTick: {
            show: false
        },
        splitLine: {
            lineStyle: {
                color: '#E4EBF057',
            }
        },
        axisLabel: {
            fontSize: 11
        }
    },
    legend: {
        // left: 45,
        // top: -3,
        x: 'center',
        y: 'bottom',
        itemWidth: 16,
        itemHeight: 3,
        borderRadius: 2,
        icon: 'roundRect',
        padding: [15, 0],
        textStyle: {
            color: '#000000a6'
        }
    },
    tooltip: {
        backgroundColor: "#ffffffe8",
        textStyle: {
            color: "#000000e6",
        },
        extraCssText: "box-shadow: 0px 1px 7px #00000017;",
        appendToBody: true,
        confine: true,
    },
};
/** @type {?} */
export const chartsConsts = {
    /**
     * series label 字体大小
     */
    seriesLabelFontSize: 9,
};
const ɵ0 = /**
 * @param {?} c
 * @return {?}
 */
c => c.split(',');
/** @type {?} */
const splitedGradientColors = PlThemeColors.gradient.map((ɵ0));
/** @type {?} */
export const getChartColorsPure = (/**
 * @param {?} params
 * @return {?}
 */
(params) => {
    return PlThemeColors.pure[params.dataIndex];
});
/** @type {?} */
export const getChartColorsGradient = (/**
 * @param {?} params
 * @return {?}
 */
(params) => {
    return new LinearGradient(0, 0, 0, 1, [
        { offset: 0, color: splitedGradientColors[params.dataIndex % splitedGradientColors.length][0] },
        { offset: 1, color: splitedGradientColors[params.dataIndex % splitedGradientColors.length][1] }
    ]);
});
/** @type {?} */
export const getChartColorsGradientOne = (/**
 * @param {?} params
 * @return {?}
 */
(params) => {
    return new LinearGradient(1, 0, 0, 0, [
        { offset: 0, color: splitedGradientColors[params.dataIndex % splitedGradientColors.length][0] },
        { offset: 1, color: splitedGradientColors[params.dataIndex % splitedGradientColors.length][1] }
    ]);
})
/**
 * 对于有 “y 轴自定义范围” 的 charts 元部件，data 都要用它过滤后再用
 * @param obj 元部件实例
 */
;
/**
 * 对于有 “y 轴自定义范围” 的 charts 元部件，data 都要用它过滤后再用
 * @param {?} obj 元部件实例
 * @param {?} data
 * @return {?}
 */
export function filterYData(obj, data) {
    if (obj.filterYFlag) {
        /** @type {?} */
        const yFields = obj.yFieldObjs.map((/**
         * @param {?} o
         * @return {?}
         */
        o => o.valueField.name));
        return data.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            return !yFields.some((/**
             * @param {?} yField
             * @return {?}
             */
            yField => item[yField] <= obj.minY || item[yField] >= obj.maxY));
        }));
    }
    return data;
}
/** @type {?} */
export const EchartsPropertyOptions = {
    showRefresh: {
        category: 'data',
        displayName: 'showRefreshBtn',
        type: PropertyTypes.Bool,
        boolOption: {
            default: true
        }
    },
    filterYFlag: {
        type: PropertyTypes.Bool,
        category: 'data',
        displayName: 'setYRange'
    },
    minY: {
        type: PropertyTypes.Number,
        category: 'data',
        displayName: 'minY',
        numberOption: {
            default: 0
        },
        /**
         * @param {?} values
         * @return {?}
         */
        ifShow(values) { return values.filterYFlag; }
    },
    maxY: {
        type: PropertyTypes.Number,
        category: 'data',
        displayName: 'maxY',
        numberOption: {
            default: 1000
        },
        /**
         * @param {?} values
         * @return {?}
         */
        ifShow(values) { return values.filterYFlag; }
    },
    showLabel: {
        category: 'style',
        type: PropertyTypes.Bool,
        displayName: 'showNumber'
    },
    showTooltip: {
        category: 'style',
        type: PropertyTypes.Bool,
        boolOption: { default: true },
        displayName: 'showHoverDetail'
    },
    legendStyleOption: {
        type: PropertyTypes.Enum,
        displayName: 'legendType',
        category: 'style',
        enumOption: {
            default: 'scroll',
            items: [
                { value: 'none', displayName: 'notVisible' },
                { value: 'plain', displayName: 'showAll' },
                { value: 'scroll', displayName: 'scrollOverflow' }
            ]
        }
    },
    valueStyleOption: {
        type: PropertyTypes.Enum,
        displayName: 'valueType',
        category: 'style',
        enumOption: {
            default: 'right',
            items: [
                { value: 'right', displayName: 'right' },
                { value: 'insideRight', displayName: 'inside' },
            ]
        }
    }
}
/**
 * 三种传参方式：
 * 1. '#000', '#fff'
 * 2. '#000,#fff'
 * 3. '#000'
 */
;
/**
 * 三种传参方式：
 * 1. '#000', '#fff'
 * 2. '#000,#fff'
 * 3. '#000'
 * @param {?} c1
 * @param {?=} c2
 * @return {?}
 */
export function echartsGradient(c1, c2) {
    // 若只传了一个值、还不是逗号分割，则直接返回此纯色
    if (!c2 && !c1.includes(','))
        return c1;
    // 若只传了一个值、是逗号分割，就分割为俩颜色
    if (!c2)
        [c1, c2] = c1.split(',');
    return new LinearGradient(0, 0, 0, 1, [
        { offset: 0, color: c1 },
        { offset: 1, color: c2 }
    ]);
}
// TODO: 暂时给资金做国际化
/** @type {?} */
export const tempMap = {
    '金额': {
        'en': 'Amount',
        'zh-CHS': '金额',
    },
    '数量': {
        'en': 'Number',
        'zh-CHS': '数量',
    }
}
// TODO: 暂时给资金国际化
;
// TODO: 暂时给资金国际化
/**
 * @param {?} dataField
 * @param {?} currentLang
 * @return {?}
 */
export function tempLabel(dataField, currentLang) {
    const { label } = dataField;
    return tempMap[label] && tempMap[label][currentLang]
        ? tempMap[label][currentLang]
        : dataField.label;
}
export { ɵ0 };
