// @gspwidget/widget-devkit/esm2015/lib/preset/preset.jump.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PropertyTypes, HelpStructure } from '../../types';
import { map } from 'rxjs/operators';
import { list2Tree } from '@gspwidget/util';
/**
 * 某一条数据联查跳转的配置属性的类型
 * @record
 */
export function DataJumpConfig() { }
if (false) {
    /** @type {?} */
    DataJumpConfig.prototype.jumpType;
    /**
     * 菜单
     * @type {?|undefined}
     */
    DataJumpConfig.prototype.funcIdSource;
    /**
     * funcId from data
     * @type {?|undefined}
     */
    DataJumpConfig.prototype.funcIdDataField;
    /**
     * funcId fixed
     * @type {?|undefined}
     */
    DataJumpConfig.prototype.funcId;
    /** @type {?|undefined} */
    DataJumpConfig.prototype.funcParams;
    /**
     * 新标签页
     * @type {?|undefined}
     */
    DataJumpConfig.prototype.url;
    /** @type {?|undefined} */
    DataJumpConfig.prototype.urlParams;
    /** @type {?|undefined} */
    DataJumpConfig.prototype.queryParams;
}
/**
 * 全局联查跳转的配置属性的类型
 * @record
 */
export function DatalessJumpConfig() { }
if (false) {
    /** @type {?} */
    DatalessJumpConfig.prototype.jumpType;
    /**
     * 菜单
     * @type {?|undefined}
     */
    DatalessJumpConfig.prototype.funcId;
    /** @type {?|undefined} */
    DatalessJumpConfig.prototype.funcParams;
    /**
     * 新标签页
     * @type {?|undefined}
     */
    DatalessJumpConfig.prototype.url;
}
/** @enum {string} */
const JumpTypes = {
    None: '0',
    Func: '1',
    ChartTable: '2',
    NewBrowserTab: '3',
};
export { JumpTypes };
/**
 * 选择联查类型
 * @type {?}
 */
export const jumpTypeOption = {
    type: PropertyTypes.Enum,
    enumOption: {
        default: JumpTypes.None,
        items: [
            { value: JumpTypes.None, displayName: 'none' },
            { value: JumpTypes.Func, displayName: 'toFuncOrInvoke' },
            // { value: JumpTypes.ChartTable, displayName: '跳转至预制图表页' },
            { value: JumpTypes.NewBrowserTab, displayName: 'browserNewTab' }
        ]
    }
}
// let allFuncs
/** 菜单/联查帮助 option */
;
// let allFuncs
/**
 * 菜单/联查帮助 option
 * @type {?}
 */
export const funcOption = {
    type: PropertyTypes.Help,
    tip: '选菜单/联查的帮助',
    helpOption: {
        getData: (/**
         * @param {?} httpClient
         * @return {?}
         */
        (httpClient) => {
            return httpClient
                .get('/api/runtime/sys/v1.0/functions/allFuncs')
                .pipe(map((/**
             * @param {?} funcs
             * @return {?}
             */
            funcs => {
                // allFuncs = funcs
                return list2Tree(funcs, {
                    rootParentIdValue: '0',
                    /**
                     * @param {?} item
                     * @return {?}
                     */
                    beforeProcess(item) {
                        if (item.layer === '3') {
                            item.isLeaf = true;
                            item.isTableParent = true;
                        }
                    }
                });
            })));
        }),
        structure: HelpStructure.treeTable,
        tableFields: ['code', 'name'],
        valueField: 'id',
        nameField: 'name'
    }
}
/** 不取决于某条数据值的联查 */
;
/**
 * 不取决于某条数据值的联查
 * @type {?}
 */
export const dataJumpOption = {
    category: 'clickJump',
    type: PropertyTypes.Object,
    objectOption: {
        objPropertyOptions: [
            Object.assign({ name: 'jumpType', displayName: 'jumpType' }, jumpTypeOption),
            /** 菜单/联查 */
            {
                name: 'funcIdSource',
                type: PropertyTypes.Enum,
                enumOption: {
                    items: ['fixed', 'fromData']
                },
                /**
                 * @param {?} c
                 * @return {?}
                 */
                ifShow(c) { return c.jumpType === JumpTypes.Func; }
            }, {
                name: 'funcIdDataField',
                type: PropertyTypes.DataField,
                /**
                 * @param {?} c
                 * @return {?}
                 */
                ifShow(c) {
                    return c.jumpType === JumpTypes.Func && c.funcIdSource === 'fromData';
                }
            },
            Object.assign({ name: 'funcId', displayName: 'func/invoke', /**
                 * @param {?} c
                 * @return {?}
                 */
                ifShow(c) {
                    return c.jumpType === JumpTypes.Func && c.funcIdSource === 'fixed';
                } }, funcOption),
            {
                name: 'funcParams',
                displayName: 'routeParams',
                type: PropertyTypes.Object,
                isArray: true,
                objectOption: {
                    objPropertyOptions: [{
                            name: 'key',
                            displayName: 'paramKey',
                            textOption: { i18n: false },
                            type: PropertyTypes.Text,
                        }, {
                            name: 'valueType',
                            displayName: 'valueType',
                            type: PropertyTypes.Enum,
                            enumOption: {
                                items: [
                                    { value: 'fieldName', displayName: 'dataFieldValue' },
                                    { value: 'fixedValue', displayName: 'fixedValue' }
                                ]
                            }
                        }, {
                            name: 'fixedValue',
                            displayName: 'paramFixedValue',
                            type: PropertyTypes.Text,
                            textOption: { i18n: false },
                            /**
                             * @param {?} funcParam
                             * @return {?}
                             */
                            ifShow(funcParam) { return funcParam.valueType === 'fixedValue'; }
                        }, {
                            name: 'fieldName',
                            displayName: 'paramValueField',
                            type: PropertyTypes.DataField,
                            /**
                             * @param {?} funcParam
                             * @return {?}
                             */
                            ifShow(funcParam) { return funcParam.valueType === 'fieldName'; }
                        }]
                },
                /**
                 * @param {?} values
                 * @return {?}
                 */
                ifShow(values) { return values.jumpType === JumpTypes.Func; }
            },
            /** 浏览器标签 */
            {
                name: 'url',
                displayName: 'newTabUrl',
                type: PropertyTypes.Text,
                textOption: { i18n: false },
                /**
                 * @param {?} values
                 * @return {?}
                 */
                ifShow(values) { return values.jumpType === JumpTypes.NewBrowserTab; }
            }, {
                name: 'urlParams',
                displayName: 'urlParams',
                type: PropertyTypes.Object,
                isArray: true,
                objectOption: {
                    objPropertyOptions: [{
                            name: 'valueType',
                            displayName: 'valueType',
                            type: PropertyTypes.Enum,
                            enumOption: {
                                items: [
                                    { value: 'fieldName', displayName: 'dataFieldValue' },
                                    { value: 'fixedValue', displayName: 'fixedValue' }
                                ]
                            }
                        }, {
                            name: 'fixedValue',
                            displayName: 'paramFixedValue',
                            type: PropertyTypes.Text,
                            textOption: { i18n: false },
                            /**
                             * @param {?} urlParam
                             * @return {?}
                             */
                            ifShow(urlParam) { return urlParam.valueType === 'fixedValue'; }
                        }, {
                            name: 'fieldName',
                            displayName: 'paramValueField',
                            type: PropertyTypes.DataField,
                            /**
                             * @param {?} urlParam
                             * @return {?}
                             */
                            ifShow(urlParam) { return urlParam.valueType === 'fieldName'; }
                        }]
                },
                /**
                 * @param {?} values
                 * @return {?}
                 */
                ifShow(values) { return values.jumpType === JumpTypes.NewBrowserTab; }
            }, {
                name: 'queryParams',
                displayName: 'urlQueryParams',
                type: PropertyTypes.Object,
                isArray: true,
                objectOption: {
                    objPropertyOptions: [{
                            name: 'key',
                            displayName: 'paramKey',
                            type: PropertyTypes.Text,
                            textOption: { i18n: false },
                        }, {
                            name: 'valueType',
                            displayName: 'valueType',
                            type: PropertyTypes.Enum,
                            enumOption: {
                                items: [
                                    { value: 'fieldName', displayName: 'dataFieldValue' },
                                    { value: 'fixedValue', displayName: 'fixedValue' }
                                ]
                            }
                        }, {
                            name: 'fixedValue',
                            displayName: 'paramFixedValue',
                            type: PropertyTypes.Text,
                            textOption: { i18n: false },
                            /**
                             * @param {?} funcParam
                             * @return {?}
                             */
                            ifShow(funcParam) { return funcParam.valueType === 'fixedValue'; }
                        }, {
                            name: 'fieldName',
                            displayName: 'paramValueField',
                            type: PropertyTypes.DataField,
                            /**
                             * @param {?} funcParam
                             * @return {?}
                             */
                            ifShow(funcParam) { return funcParam.valueType === 'fieldName'; }
                        }]
                },
                /**
                 * @param {?} values
                 * @return {?}
                 */
                ifShow(values) { return values.jumpType === JumpTypes.NewBrowserTab; }
            }
        ]
    }
}
/** 决于某条数据值的联查 */
;
/**
 * 决于某条数据值的联查
 * @type {?}
 */
export const datalessJumpOption = {
    category: 'clickJump',
    type: PropertyTypes.Object,
    objectOption: {
        objPropertyOptions: [
            Object.assign({ name: 'jumpType', displayName: 'jumpType' }, jumpTypeOption),
            Object.assign({ name: 'funcId', displayName: 'func/invoke', /**
                 * @param {?} values
                 * @return {?}
                 */
                ifShow(values) { return values.jumpType === JumpTypes.Func; } }, funcOption),
            {
                name: 'funcParams',
                displayName: 'routeParams',
                type: PropertyTypes.Object,
                isArray: true,
                objectOption: {
                    objPropertyOptions: [{
                            name: 'key',
                            displayName: 'paramKey',
                            type: PropertyTypes.Text,
                            textOption: { i18n: false },
                        }, {
                            name: 'value',
                            displayName: 'paramValue',
                            type: PropertyTypes.Text,
                            textOption: { i18n: false },
                        }]
                },
                /**
                 * @param {?} values
                 * @return {?}
                 */
                ifShow(values) { return values.jumpType === JumpTypes.Func; }
            },
            /** 浏览器标签 */
            {
                name: 'url',
                displayName: 'newTabUrl',
                type: PropertyTypes.Text,
                textOption: { i18n: false },
                /**
                 * @param {?} values
                 * @return {?}
                 */
                ifShow(values) { return values.jumpType === JumpTypes.NewBrowserTab; }
            }
        ]
    }
}
/** @deprecated 请用 datalessJump 或 dataJump */
;
/**
 * @deprecated 请用 datalessJump 或 dataJump
 * @type {?}
 */
export const CommonPropertyOptions = {
    jump: datalessJumpOption
};
