// @gspwidget/widget-devkit/esm2015/lib/preset/zrender.linear-gradient.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import Gradient from "./zrender.gradient";
/**
 * @record
 */
export function LinearGradientObject() { }
if (false) {
    /** @type {?} */
    LinearGradientObject.prototype.type;
    /** @type {?} */
    LinearGradientObject.prototype.x;
    /** @type {?} */
    LinearGradientObject.prototype.y;
    /** @type {?} */
    LinearGradientObject.prototype.x2;
    /** @type {?} */
    LinearGradientObject.prototype.y2;
    /** @type {?} */
    LinearGradientObject.prototype.global;
}
/**
 * x, y, x2, y2 are all percent from 0 to 1 when globalCoord is false
 */
export default class LinearGradient extends Gradient {
    /**
     * @param {?} x
     * @param {?} y
     * @param {?} x2
     * @param {?} y2
     * @param {?=} colorStops
     * @param {?=} globalCoord
     */
    constructor(x, y, x2, y2, colorStops, globalCoord) {
        super(colorStops);
        // Should do nothing more in this constructor. Because gradient can be
        // declard by `color: {type: 'linear', colorStops: ...}`, where
        // this constructor will not be called.
        this.x = x == null ? 0 : x;
        this.y = y == null ? 0 : y;
        this.x2 = x2 == null ? 1 : x2;
        this.y2 = y2 == null ? 0 : y2;
        // Can be cloned
        this.type = "linear";
        // If use global coord
        this.global = globalCoord || false;
    }
}
if (false) {
    /** @type {?} */
    LinearGradient.prototype.type;
    /** @type {?} */
    LinearGradient.prototype.x;
    /** @type {?} */
    LinearGradient.prototype.y;
    /** @type {?} */
    LinearGradient.prototype.x2;
    /** @type {?} */
    LinearGradient.prototype.y2;
    /** @type {?} */
    LinearGradient.prototype.global;
}
