// @gsp-cmp/chgdr/esm5/lib/chgdr-data-viewer/chgdr-tree-builder.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/chgdr-data-viewer/chgdr-tree-builder.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Guid } from '../util/guid.util';
// 参考： Farris TreeTable的节点结构
// export interface TreeNode {
//     id?: string;
//     data: any;
//     /** 叶子节点图标 */
//     icon?: any;
//     /** 节点展开图标 */
//     expandedIcon?: any;
//     /** 节点折叠图标 */
//     collapsedIcon?: any;
//     /** 子节点 */
//     children?: TreeNode[];
//     /** 是否为叶子节点 */
//     leaf?: boolean;
//     /** 是否展开 */
//     expanded?: boolean;
//     type?: string;
//     /** 父节点 */
//     parent?: TreeNode;
//     /** 是否允许选中 */
//     selectable?: boolean;
//  }
var 
// 参考： Farris TreeTable的节点结构
// export interface TreeNode {
//     id?: string;
//     data: any;
//     /** 叶子节点图标 */
//     icon?: any;
//     /** 节点展开图标 */
//     expandedIcon?: any;
//     /** 节点折叠图标 */
//     collapsedIcon?: any;
//     /** 子节点 */
//     children?: TreeNode[];
//     /** 是否为叶子节点 */
//     leaf?: boolean;
//     /** 是否展开 */
//     expanded?: boolean;
//     type?: string;
//     /** 父节点 */
//     parent?: TreeNode;
//     /** 是否允许选中 */
//     selectable?: boolean;
//  }
ChgdrTreeNode = /** @class */ (function () {
    function ChgdrTreeNode() {
    }
    return ChgdrTreeNode;
}());
// 参考： Farris TreeTable的节点结构
// export interface TreeNode {
//     id?: string;
//     data: any;
//     /** 叶子节点图标 */
//     icon?: any;
//     /** 节点展开图标 */
//     expandedIcon?: any;
//     /** 节点折叠图标 */
//     collapsedIcon?: any;
//     /** 子节点 */
//     children?: TreeNode[];
//     /** 是否为叶子节点 */
//     leaf?: boolean;
//     /** 是否展开 */
//     expanded?: boolean;
//     type?: string;
//     /** 父节点 */
//     parent?: TreeNode;
//     /** 是否允许选中 */
//     selectable?: boolean;
//  }
export { ChgdrTreeNode };
if (false) {
    /**
     * 节点ID，用于内部处理，请勿使用
     * @type {?}
     */
    ChgdrTreeNode.prototype.id;
    /**
     * 节点名称
     * @type {?}
     */
    ChgdrTreeNode.prototype.name;
    /**
     * 父节点
     * @type {?}
     */
    ChgdrTreeNode.prototype.parent;
    /**
     * 子节点列表
     * @type {?}
     */
    ChgdrTreeNode.prototype.children;
    /**
     * 是否展开
     * @type {?}
     */
    ChgdrTreeNode.prototype.expanded;
    /**
     * 是否隐藏
     * @type {?}
     */
    ChgdrTreeNode.prototype.hidden;
    /**
     * 展示树结构用的数据实体
     * @type {?}
     */
    ChgdrTreeNode.prototype.data;
    /** @type {?} */
    ChgdrTreeNode.prototype.chgdrRow;
    /** @type {?} */
    ChgdrTreeNode.prototype.chgdrData;
    /** @type {?} */
    ChgdrTreeNode.prototype.isRoot;
}
var ChgdrTreeBuilder = /** @class */ (function () {
    function ChgdrTreeBuilder() {
    }
    /**
     * @param {?} chgdrData
     * @return {?}
     */
    ChgdrTreeBuilder.prototype.build = /**
     * @param {?} chgdrData
     * @return {?}
     */
    function (chgdrData) {
        this.chgdrData = chgdrData;
        if (chgdrData.entityNodes.length == 0) {
            console.warn("无变更日志");
            return [];
        }
        /** @type {?} */
        var rootEntityNode = chgdrData.entityNodes[0];
        /** @type {?} */
        var rootRows = rootEntityNode.rows;
        if (rootRows.length == 0) {
            console.warn("无根变更日志行");
            return [];
        }
        /** @type {?} */
        var rootRow = rootRows[0];
        /** @type {?} */
        var node = this.rowToNode(rootRow);
        node.isRoot = true;
        node.expanded = true;
        return [node];
    };
    /**
     * @private
     * @param {?} row
     * @return {?}
     */
    ChgdrTreeBuilder.prototype.rowToNode = /**
     * @private
     * @param {?} row
     * @return {?}
     */
    function (row) {
        var _this = this;
        /** @type {?} */
        var node = new ChgdrTreeNode();
        node.id = Guid.newGuid();
        node.name = this.buildTreeNodeName(row);
        node.data = {
            id: node.id,
            name: node.name
        };
        node.chgdrRow = row;
        node.chgdrData = this.chgdrData;
        node.children = [];
        row.entityNodes.forEach((/**
         * @param {?} entityNode
         * @return {?}
         */
        function (entityNode) {
            entityNode.rows.forEach((/**
             * @param {?} childRow
             * @return {?}
             */
            function (childRow) {
                /** @type {?} */
                var childNode = _this.rowToNode(childRow);
                childNode.parent = node;
                node.children.push(childNode);
            }));
        }));
        return node;
    };
    /**
     * @private
     * @param {?} row
     * @return {?}
     */
    ChgdrTreeBuilder.prototype.buildTreeNodeName = /**
     * @private
     * @param {?} row
     * @return {?}
     */
    function (row) {
        /** @type {?} */
        var dataCodeName;
        if (row.dataCodes && row.dataCodes.length > 0) {
            dataCodeName = row.dataCodes[0].fieldValue;
        }
        if (dataCodeName) {
            return row.entityName + "-" + dataCodeName + "【" + row.operateType.name + "】";
        }
        else {
            return row.entityName + "【" + row.operateType.name + "】";
        }
    };
    return ChgdrTreeBuilder;
}());
export { ChgdrTreeBuilder };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ChgdrTreeBuilder.prototype.chgdrData;
}
