// @gsp-cmp/chgdr/esm5/lib/chgdr-data-viewer/chgdr-data.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/chgdr-data-viewer/chgdr-data.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 基本信息
 */
var /**
 * 基本信息
 */
ChgdrData = /** @class */ (function () {
    function ChgdrData() {
    }
    return ChgdrData;
}());
/**
 * 基本信息
 */
export { ChgdrData };
if (false) {
    /** @type {?} */
    ChgdrData.prototype.entityName;
    /** @type {?} */
    ChgdrData.prototype.userName;
    /** @type {?} */
    ChgdrData.prototype.operateType;
    /** @type {?} */
    ChgdrData.prototype.reason;
    /** @type {?} */
    ChgdrData.prototype.dataId;
    /** @type {?} */
    ChgdrData.prototype.dataCode;
    /** @type {?} */
    ChgdrData.prototype.dataCodes;
    /** @type {?} */
    ChgdrData.prototype.changeTime;
    /**
     * 本变更日志的根节点
     * @type {?}
     */
    ChgdrData.prototype.entityNodes;
}
/**
 * 发生变更的实体节点
 */
var /**
 * 发生变更的实体节点
 */
ChgdrDataEntityNode = /** @class */ (function () {
    function ChgdrDataEntityNode() {
    }
    return ChgdrDataEntityNode;
}());
/**
 * 发生变更的实体节点
 */
export { ChgdrDataEntityNode };
if (false) {
    /** @type {?} */
    ChgdrDataEntityNode.prototype.entityCode;
    /** @type {?} */
    ChgdrDataEntityNode.prototype.entityName;
    /** @type {?} */
    ChgdrDataEntityNode.prototype.rows;
}
/**
 * 发生变更的行
 */
var /**
 * 发生变更的行
 */
ChgdrDataRowNode = /** @class */ (function () {
    function ChgdrDataRowNode() {
    }
    return ChgdrDataRowNode;
}());
/**
 * 发生变更的行
 */
export { ChgdrDataRowNode };
if (false) {
    /** @type {?} */
    ChgdrDataRowNode.prototype.id;
    /** @type {?} */
    ChgdrDataRowNode.prototype.entityCode;
    /** @type {?} */
    ChgdrDataRowNode.prototype.entityName;
    /** @type {?} */
    ChgdrDataRowNode.prototype.parentDataId;
    /** @type {?} */
    ChgdrDataRowNode.prototype.dataId;
    /** @type {?} */
    ChgdrDataRowNode.prototype.dataCode;
    /** @type {?} */
    ChgdrDataRowNode.prototype.dataCodes;
    /** @type {?} */
    ChgdrDataRowNode.prototype.operateType;
    /** @type {?} */
    ChgdrDataRowNode.prototype.changes;
    /**
     * 发生变更的行的下级实体
     * @type {?}
     */
    ChgdrDataRowNode.prototype.entityNodes;
}
var ChangeColumnItem = /** @class */ (function () {
    function ChangeColumnItem() {
        /**
         * 如果当前字段为关联字段，则属性为当前字段的关联带出字段
         */
        this.children = [];
    }
    return ChangeColumnItem;
}());
export { ChangeColumnItem };
if (false) {
    /** @type {?} */
    ChangeColumnItem.prototype.fieldLabel;
    /** @type {?} */
    ChangeColumnItem.prototype.fieldName;
    /** @type {?} */
    ChangeColumnItem.prototype.oldValue;
    /** @type {?} */
    ChangeColumnItem.prototype.newValue;
    /**
     * 是否关联字段
     * @type {?}
     */
    ChangeColumnItem.prototype.hasAssociation;
    /**
     * 原始关联字段
     * @type {?}
     */
    ChangeColumnItem.prototype.originalColumnItem;
    /**
     * 是否关联带出字段
     * @type {?}
     */
    ChangeColumnItem.prototype.isAssociationRefField;
    /**
     * 关联带出字段所属的字段
     * @type {?}
     */
    ChangeColumnItem.prototype.belongFieldLabelId;
    /**
     * 如果当前字段为关联字段，则属性为当前字段的关联带出字段
     * @type {?}
     */
    ChangeColumnItem.prototype.children;
}
var DataCodeField = /** @class */ (function () {
    function DataCodeField() {
    }
    return DataCodeField;
}());
export { DataCodeField };
if (false) {
    /** @type {?} */
    DataCodeField.prototype.fieldLabelId;
    /** @type {?} */
    DataCodeField.prototype.fieldName;
    /** @type {?} */
    DataCodeField.prototype.fieldValue;
}
