// @farris/ui-filter/esm2015/lib/entity/list-filter.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class listFilterItem {
    /**
     * @param {?} _id
     * @param {?} _groupName
     */
    constructor(_id, _groupName) {
        this.id = _id;
        this.name = _groupName;
        this.width = 0;
        this.isExtend = false;
    }
    // 设置宽度
    /**
     * @param {?} value
     * @return {?}
     */
    setWidth(value) {
        this.width = parseInt(value, 10);
    }
    // 获取宽度
    /**
     * @return {?}
     */
    getWidth() {
        return this.width;
    }
}
if (false) {
    /** @type {?} */
    listFilterItem.prototype.id;
    /** @type {?} */
    listFilterItem.prototype.name;
    /** @type {?} */
    listFilterItem.prototype.width;
    /** @type {?} */
    listFilterItem.prototype.isExtend;
}
export class listFilterGroup {
    constructor() {
        this.responsedIndex = 0;
        this.listItems = [];
    }
}
if (false) {
    /** @type {?} */
    listFilterGroup.prototype.listItems;
    /** @type {?} */
    listFilterGroup.prototype.responsedIndex;
}
