// @farris/ui-organization-selector/esm2015/utils/api.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/api.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { of } from 'rxjs';
export class ApiService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * 获取行政组织或系统用户
     * @param {?} param
     * @return {?} observable
     */
    getOrganization(param) {
        /** @type {?} */
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        return this.http.post(this.absUrl + '/org', param, options);
    }
    /**
     * @param {?} absOrgType
     * @return {?}
     */
    getOrgGroups(absOrgType) {
        /** @type {?} */
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        return this.http.get(this.groupApiUrl + '/' + absOrgType, options);
    }
    /**
     * @param {?} groupId
     * @param {?} absOrgType
     * @return {?}
     */
    getGroupMembers(groupId, absOrgType) {
        /** @type {?} */
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        if (!groupId) {
            return of(null);
        }
        /** @type {?} */
        const url = this.groupApiUrl + '/' + groupId + '/absOrgType/' + absOrgType;
        return this.http.get(url, options);
    }
    /**
     * @param {?} groupInfo
     * @return {?}
     */
    saveGroup(groupInfo) {
        /** @type {?} */
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        return this.http.post(this.groupApiUrl, groupInfo, options);
    }
}
ApiService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ApiService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /** @type {?} */
    ApiService.prototype.absUrl;
    /** @type {?} */
    ApiService.prototype.groupApiUrl;
    /**
     * @type {?}
     * @private
     */
    ApiService.prototype.http;
}
