// @farris/ui-organization-selector/esm2015/utils/index.type.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/index.type.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function OrgRequestParam() { }
if (false) {
    /** @type {?} */
    OrgRequestParam.prototype.absOrgType;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.layer;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.codeOrName;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.name;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.parentID;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.isSearchTree;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getCurUser;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.notGetStopOrg;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.id;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.idList;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getAllParentOrg;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getOrgAllPath;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getOrgAllPathInfo;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.withAllSub;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.layerList;
}
/** @enum {string} */
const ChangeType = {
    Append: "Append",
    Remove: "Remove",
    ValueChanged: "ValueChanged",
};
export { ChangeType };
/**
 * @record
 */
export function Param() { }
if (false) {
    /** @type {?} */
    Param.prototype.value;
    /** @type {?} */
    Param.prototype.changeType;
    /** @type {?} */
    Param.prototype.from;
}
/**
 * @record
 */
export function Organization() { }
if (false) {
    /** @type {?} */
    Organization.prototype.orgId;
    /** @type {?} */
    Organization.prototype.code;
    /** @type {?} */
    Organization.prototype.name;
    /** @type {?} */
    Organization.prototype.orgAllPath;
    /** @type {?} */
    Organization.prototype.path;
    /** @type {?} */
    Organization.prototype.layer;
    /** @type {?|undefined} */
    Organization.prototype.checked;
    /** @type {?|undefined} */
    Organization.prototype.indexId;
    /** @type {?|undefined} */
    Organization.prototype.enable;
}
/**
 * @record
 */
export function Group() { }
if (false) {
    /** @type {?} */
    Group.prototype.public;
    /** @type {?} */
    Group.prototype.personal;
}
/**
 * @record
 */
export function GroupItem() { }
if (false) {
    /** @type {?} */
    GroupItem.prototype.id;
    /** @type {?} */
    GroupItem.prototype.title;
    /** @type {?|undefined} */
    GroupItem.prototype.index;
    /** @type {?|undefined} */
    GroupItem.prototype.description;
    /** @type {?} */
    GroupItem.prototype.total;
    /** @type {?|undefined} */
    GroupItem.prototype.avatar;
    /** @type {?|undefined} */
    GroupItem.prototype.checked;
}
/**
 * @record
 */
export function SaveGroupReqeustParam() { }
if (false) {
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.groupId;
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.groupCode;
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.title;
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.description;
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.absOrgType;
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.addIds;
    /** @type {?|undefined} */
    SaveGroupReqeustParam.prototype.deleteIds;
    /** @type {?} */
    SaveGroupReqeustParam.prototype.changeType;
}
