// @farris/ui-organization-selector/esm2015/utils/locale.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/locale.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const ORGANIZATION_SELECTOR_LOCALE = {
    "zh-CHS": {
        organizationSelector: {
            selection: {
                selectionDialogTitle: '组织选择',
                organization: '组织架构',
                group: '我的群组',
                favorites: '我的收藏',
                placeHolder: '请选择',
                searchPlaceHolder: '请输入编号或名称搜索',
                confirm: '确定',
                cancel: '取消',
                firstSelectionText: '已选',
                lastSelectionText: ' 个',
                clear: '清空已选'
            },
            tabCommon: {
                firstCheckAllText: '全选（',
                lastCheckAllText: '）',
                noOrgData: '暂无数据',
                more: '更多',
                noMore: '暂无已选组织'
            },
            organization: {
                orgCode: '组织编号',
                orgName: '组织名称',
                favorite: '收藏'
            },
            group: {
                newGroupName: '群组名称',
                newGroupNote: '群组备注',
                newGroupPlaceHolder: '请输入',
                publicGroup: '公共群组',
                privateGroup: '个人群组',
                appendMembers: '添加组织',
                items: '个组织',
                saveGroup: '保存为群组',
                editGroup: '编辑群组',
                checkAll: '全选'
            },
            searchPage: {
                firstResultText: '搜索结果（',
                lastResultText: '）',
                close: '关闭'
            },
            successMessage: {
                saveGroup: '群组保存成功！',
                editGroup: '修改群组信息成功！',
                deleteGroup: '删除群组成功！',
                appendOrganization: '添加组织成功！',
                deleteOrganization: '删除组织成功！',
            },
            errorMessage: {
                mapFields: '字段映射中未添加组织唯一标识。',
                multiSelect: '单选模式只允许选中一条数据，请检查！',
                limit: '群组内暂不支持保存超过1000个组织，请修改。',
                saveGroup: '群组保存失败！',
                editGroup: '修改群组信息失败！',
                deleteGroup: '删除群组失败！',
                getGroup: '获取群组数据失败！',
                saveOrganization: '群组内组织保存失败！',
                appendOrganization: '添加组织失败！',
                deleteOrganization: '删除组织失败！',
                getOrganization: '获取群组内组织数据失败！'
            },
            confirmMessage: {
                deleteGroup: '是否删除群组：',
                deleteOrganization: '是否删除组织：',
                mark: '？'
            }
        }
    },
    "zh-CHT": {
        organizationSelector: {
            selection: {
                selectionDialogTitle: '組織選擇',
                organization: '組織架構',
                group: '我的群組',
                favorites: '我的收藏',
                placeHolder: '請選擇',
                searchPlaceHolder: '請輸入編號或名稱搜索',
                confirm: '確定',
                cancel: '取消',
                firstSelectionText: '已選',
                lastSelectionText: ' 個',
                clear: '清空已選'
            },
            tabCommon: {
                firstCheckAllText: '全選（',
                lastCheckAllText: '）',
                noOrgData: '暫無數據',
                more: '更多',
                noMore: '暫無已選組織'
            },
            organization: {
                orgCode: '組織編號',
                orgName: '組織名稱',
                favorite: '收藏'
            },
            group: {
                newGroupName: '群組名稱',
                newGroupNote: '群組備注',
                newGroupPlaceHolder: '請輸入',
                publicGroup: '公共群組',
                privateGroup: '個人群組',
                appendMembers: '添加組織',
                items: '個組織',
                saveGroup: '保存爲群組',
                editGroup: '編輯群組',
                checkAll: '全選'
            },
            searchPage: {
                firstResultText: '搜索結果（',
                lastResultText: '）',
                close: '關閉'
            },
            successMessage: {
                saveGroup: '群組保存成功！',
                editGroup: '修改群組信息成功！',
                deleteGroup: '刪除群組成功！',
                appendOrganization: '添加組織成功！',
                deleteOrganization: '刪除組織成功！',
            },
            errorMessage: {
                mapFields: '字段映射中未添加組織唯一標識。',
                multiSelect: '單選模式只允許選中一條數據，請檢查！',
                limit: '群組内暫不支持保存超過1000個組織，請修改。',
                saveGroup: '群組保存失敗！',
                editGroup: '修改群組信息失敗！',
                deleteGroup: '刪除群組失敗！',
                getGroup: '獲取群組數據失敗！',
                saveOrganization: '群組内組織保存失敗！',
                appendOrganization: '添加組織失敗！',
                deleteOrganization: '刪除組織失敗！',
                getOrganization: '獲取群組内組織數據失敗！'
            },
            confirmMessage: {
                deleteGroup: '是否刪除群組：',
                deleteOrganization: '是否刪除組織：',
                mark: '？'
            }
        }
    },
    en: {
        organizationSelector: {
            selection: {
                selectionDialogTitle: 'Select Organization',
                organization: 'Organizations',
                group: 'Groups',
                favorites: 'Favorites',
                placeHolder: 'Please select',
                searchPlaceHolder: 'Search Organization By Code or Name',
                confirm: 'Confirm',
                cancel: 'Cancel',
                firstSelectionText: 'Total',
                lastSelectionText: ' items',
                clear: 'Clear'
            },
            tabCommon: {
                firstCheckAllText: 'All(',
                lastCheckAllText: ')',
                noOrgData: 'Empty Data',
                more: 'More Data',
                noMore: 'Empty Selected Data'
            },
            organization: {
                orgCode: 'Code',
                orgName: 'Name',
                favorite: 'Favorites'
            },
            group: {
                newGroupName: 'Name',
                newGroupNote: 'Note',
                newGroupPlaceHolder: 'Please input',
                publicGroup: 'Public Group',
                privateGroup: 'Private Group',
                appendMembers: 'Append Organizations',
                items: ' organizations',
                saveGroup: 'Save Group',
                editGroup: 'Edit Group',
                checkAll: 'All'
            },
            searchPage: {
                firstResultText: 'Search Result(',
                lastResultText: ')',
                close: 'Close'
            },
            successMessage: {
                saveGroup: 'Group Saved Successfully!',
                editGroup: 'Group Info Modified Successfully!',
                deleteGroup: 'Group Deleted Successfully!',
                appendOrganization: 'Organization Appended Successfully!',
                deleteOrganization: 'Organization Deleted Successfully!',
            },
            errorMessage: {
                mapFields: 'OrgId is not included in Mapping-Fields.',
                multiSelect: 'Only one item can be selected in single-select mode, please check! ',
                limit: 'Organizations saved in group cannot exceed 1,000 temporarily, please correct.',
                saveGroup: 'Failed to Save Group!',
                editGroup: 'Failed to Modify Group Info!',
                deleteGroup: 'Failed to Delete Group!',
                getGroup: 'Failed to Get Group Data!',
                saveOrganization: 'Failed to Save Organization!',
                appendOrganization: 'Failed to Append Organization!',
                deleteOrganization: 'Failed to Delete Organization!',
                getOrganization: 'Failed to Get Organization Data!'
            },
            confirmMessage: {
                deleteGroup: 'Delete Group: ',
                deleteOrganization: 'Delete Organization: ',
                mark: '?'
            }
        }
    }
};
