// @farris/ui-responsive/esm5/lib/container/res-container.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, HostBinding, Renderer2 } from '@angular/core';
var FarrisContainerDirective = /** @class */ (function () {
    function FarrisContainerDirective(el, render) {
        this.el = el;
        this.render = render;
        this.isContainer = true;
    }
    Object.defineProperty(FarrisContainerDirective.prototype, "width", {
        get: /**
         * @return {?}
         */
        function () {
            // 对各个单位进行处理
            return this._width;
        },
        set: /**
         * @param {?} width
         * @return {?}
         */
        function (width) {
            this._width = this.resolveSize(width);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FarrisContainerDirective.prototype, "height", {
        get: /**
         * @return {?}
         */
        function () {
            return this._height;
        },
        set: /**
         * @param {?} height
         * @return {?}
         */
        function (height) {
            this._height = this.resolveSize(height);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FarrisContainerDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        // 宽高  赋值
        this.render.setStyle(this.el.nativeElement, 'width', this.width);
        this.render.setStyle(this.el.nativeElement, 'height', this.height);
    };
    /*宽高类型是 string或者number 解析宽高  尺寸 */
    /*宽高类型是 string或者number 解析宽高  尺寸 */
    /**
     * @param {?} size
     * @return {?}
     */
    FarrisContainerDirective.prototype.resolveSize = /*宽高类型是 string或者number 解析宽高  尺寸 */
    /**
     * @param {?} size
     * @return {?}
     */
    function (size) {
        /** @type {?} */
        var regex = /px|em|rem|pt|%/;
        // 说明是字符串
        return regex.test(size) ? "" + parseInt(size, 10) + size.match(regex)[0] : size + "px";
    };
    FarrisContainerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-container]'
                },] }
    ];
    /** @nocollapse */
    FarrisContainerDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    FarrisContainerDirective.propDecorators = {
        isContainer: [{ type: HostBinding, args: ['class.container',] }],
        width: [{ type: Input }],
        height: [{ type: Input }]
    };
    return FarrisContainerDirective;
}());
export { FarrisContainerDirective };
if (false) {
    /** @type {?} */
    FarrisContainerDirective.prototype.isContainer;
    /** @type {?} */
    FarrisContainerDirective.prototype._width;
    /** @type {?} */
    FarrisContainerDirective.prototype._height;
    /**
     * @type {?}
     * @private
     */
    FarrisContainerDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FarrisContainerDirective.prototype.render;
}
