// @gsp-svc/expression/esm2015/lib/expression-event-service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * 自定义的表达式事件服务类
 * 功能：定义表达式内部的事件发布与监听接口
 */
export class ExpressionEventService {
    constructor() {
        // Observable string sources
        this.windowClosed = new Subject();
        // Observable string streams
        this.windowClosed$ = this.windowClosed.asObservable();
        //创建Observable源
        this._expressionsource = new Subject();
        //在光标处插入表达式事件发布与订阅
        this._insertexpression = new Subject();
        this._clearexpression = new Subject();
        //函数选择焦点清除
        this._clearFuncSelect = new Subject();
        //变量选择焦点清除
        this._clearVarSelect = new Subject();
    }
    // Service message commands
    /**
     * @return {?}
     */
    closeWindow() {
        this.windowClosed.next();
    }
    //定义发布表达式实体方法
    /**
     * @param {?} expressionentity
     * @return {?}
     */
    setExpression(expressionentity) {
        this._expressionsource.next(expressionentity);
    }
    //定义获得表达式实体方法
    /**
     * @return {?}
     */
    getExpression() {
        return this._expressionsource.asObservable();
    }
    /**
     * @param {?} expression
     * @return {?}
     */
    insertExpression(expression) {
        this._insertexpression.next(expression);
    }
    /**
     * @return {?}
     */
    currentExpression() {
        return this._insertexpression.asObservable();
    }
    /**
     * @return {?}
     */
    pushclearExpression() {
        this._clearexpression.next();
    }
    /**
     * @return {?}
     */
    clearExpression() {
        return this._clearexpression.asObservable();
    }
    /**
     * @return {?}
     */
    clearFuncSelect() {
        this._clearFuncSelect.next();
    }
    /**
     * @return {?}
     */
    subClearFuncSelect() {
        return this._clearFuncSelect.asObservable();
    }
    /**
     * @return {?}
     */
    clearVarSelect() {
        this._clearFuncSelect.next();
    }
    /**
     * @return {?}
     */
    subClearVarSelect() {
        return this._clearFuncSelect.asObservable();
    }
}
ExpressionEventService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ExpressionEventService.prototype.windowClosed;
    /** @type {?} */
    ExpressionEventService.prototype.windowClosed$;
    /**
     * @type {?}
     * @private
     */
    ExpressionEventService.prototype._expressionsource;
    /**
     * @type {?}
     * @private
     */
    ExpressionEventService.prototype._insertexpression;
    /**
     * @type {?}
     * @private
     */
    ExpressionEventService.prototype._clearexpression;
    /**
     * @type {?}
     * @private
     */
    ExpressionEventService.prototype._clearFuncSelect;
    /**
     * @type {?}
     * @private
     */
    ExpressionEventService.prototype._clearVarSelect;
}
