// @gsp-svc/expression/esm2015/lib/funcentity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 函数实体
 * \@property key 函数语义化名称
 * \@property params 函数参数集
 * \@property name 函数名称
 * \@property type 函数类型
 * \@property className 所在类名
 * \@property description 节点自描述
 * @author linchang 2018/12/7
 */
export class FuncEntity {
    constructor() {
        this.className = "";
    }
}
if (false) {
    /** @type {?} */
    FuncEntity.prototype.key;
    /** @type {?} */
    FuncEntity.prototype.params;
    /** @type {?} */
    FuncEntity.prototype.name;
    /** @type {?} */
    FuncEntity.prototype.type;
    /** @type {?} */
    FuncEntity.prototype.className;
    /** @type {?} */
    FuncEntity.prototype.description;
}
/**
 * 函数实体列表
 * \@property funcs 函数实体列表
 * @author linchang 2018/12/7
 */
export class FuncEntities {
}
if (false) {
    /** @type {?} */
    FuncEntities.prototype.funcs;
}
/** @enum {number} */
const FuncType = {
    "Advanced": 0,
    "Date": 1,
    "Estimate": 2,
    "String": 3,
    "Translate": 4,
    "Aggregate": 5,
    "Analytical": 6,
    "Math": 7,
    "Extend": 8 //扩展函数
    ,
};
export { FuncType };
FuncType[FuncType["Advanced"]] = "Advanced";
FuncType[FuncType["Date"]] = "Date";
FuncType[FuncType["Estimate"]] = "Estimate";
FuncType[FuncType["String"]] = "String";
FuncType[FuncType["Translate"]] = "Translate";
FuncType[FuncType["Aggregate"]] = "Aggregate";
FuncType[FuncType["Analytical"]] = "Analytical";
FuncType[FuncType["Math"]] = "Math";
FuncType[FuncType["Extend" //扩展函数
]] = "Extend" //扩展函数
;
/**
 * 参数信息
 * @param key 不同语言，统一标识
 * \@property name 语义化参数名
 * \@property type 参数类型
 * @author linchang 2018/12/7
 */
export class Parameter {
}
if (false) {
    /** @type {?} */
    Parameter.prototype.key;
    /** @type {?} */
    Parameter.prototype.name;
    /** @type {?} */
    Parameter.prototype.type;
}
/**
 * 参数信息
 * \@property key 语义化替换时脚本使用的统一主键key，一般为函数名，默认为英文
 * \@property name 脚本主键key对应的语义化字段，不同语言不同
 * \@property description 统一主键字段函数的功能描述
 * \@property type 函数类型
 * \@property keyFront 函数前缀（主要是DefaultFunction,Math）
 * \@property className 类路径（主要用户扩展函数寻找函数所在类路径）
 * @author manwenxing 2020/12/21
 */
export class FunctionEntity {
}
if (false) {
    /** @type {?} */
    FunctionEntity.prototype.key;
    /** @type {?} */
    FunctionEntity.prototype.name;
    /** @type {?} */
    FunctionEntity.prototype.params;
    /** @type {?} */
    FunctionEntity.prototype.description;
    /** @type {?} */
    FunctionEntity.prototype.type;
    /** @type {?} */
    FunctionEntity.prototype.keyFront;
    /** @type {?} */
    FunctionEntity.prototype.className;
}
/** @type {?} */
export const FunctionEntitys = [
    {
        "key": "if",
        "name": "if()\n{\n\n}",
        "params": [],
        "description": "if判断条件",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "if else",
        "name": "if()\n{\n\n}\nelse\n{\n\n}",
        "params": [],
        "description": "if else判断条件",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "for",
        "name": "循环(;;)\n{\n\n}",
        "params": [],
        "description": "for循环",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "return",
        "name": "return ;",
        "params": [],
        "description": "返回操作",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "var",
        "name": "变量",
        "params": [],
        "description": "变量定义",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "true",
        "name": "true ",
        "params": [],
        "description": "正确",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "false",
        "name": "false ",
        "params": [],
        "description": "错误",
        "type": "Advanced",
        "keyFront": "",
        "className": ""
    },
    {
        "key": "StringSubstring",
        "name": "取子字符串",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "start",
                "name": "起始位置",
                "type": "int"
            },
            {
                "key": "length",
                "name": "长度",
                "type": "int"
            }
        ],
        "description": `【功能】：在源字符串中检索子字符串，子字符串从指定的字符位置开始且具有指定的长度\n【original】：源字符串\n【start】：起始位置(从0开始)\n【length】：长度\n【返回值类型】：字符串(System.String)\n【示例】：StringSubstring("ABCDEFG",2,3)\n【返回值】："CDE"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "SubString",
        "name": "取子字符串",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "start",
                "name": "起始位置",
                "type": "int"
            },
            {
                "key": "length",
                "name": "长度",
                "type": "int"
            }
        ],
        "description": `【功能】：在源字符串中检索子字符串，子字符串从指定的字符位置开始且具有指定的长度\n【original】：源字符串\n【start】：起始位置(从0开始)\n【length】：长度\n【返回值类型】：字符串(System.String)\n【示例】：SubString("ABCDEFG",2,3)\n【返回值】："CDE"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringLength",
        "name": "取字符串长度",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：获取指定字符串的长度\n【original】：源字符串\n【返回值类型】：数值(System.Int32)\n【示例】：StringLength("语义化表达式")\n【返回值】：6`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Length",
        "name": "取字符串长度",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：获取指定字符串的长度\n【original】：源字符串\n【返回值类型】：数值(System.Int32)\n【示例】：Length("语义化表达式")\n【返回值】：6`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Contains",
        "name": "是否包含",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：判断源字符串中是否包含目标字符串\n【original】：源字符串\n【target】：目标字符串\n【返回值类型】：System.Boolean\n【示例】：Contains(salesorder.code,"表达式")\n【返回值】：true`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StartsWith",
        "name": "开头是",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：判断源字符串是否以目标字符串开头\n【original】：源字符串\n【target】：目标字符串\n【返回值类型】：System.Boolean\n【示例】：StartsWith(salesorder.code,"子单位")\n【返回值】：true`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "EndsWith",
        "name": "结尾是",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：判断源字符串是否以目标字符串结尾\n【original】：源字符串\n【target】：目标字符串\n【返回值类型】：System.Boolean\n【示例】：EndsWith(salesorder.code,\"子单位\")\n【返回值】：true`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringReplace",
        "name": "字符串替换",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "oldValue",
                "name": "被替换值",
                "type": "string"
            },
            {
                "key": "newValue",
                "name": "替换值",
                "type": "string"
            }
        ],
        "description": `【功能】：将源字符串中的"被替换值"替换为"替换值"\n【original】：源字符串\n【oldValue】：被替换值\n【newValue】：替换值\n【返回值类型】：字符串(System.String)\n【示例】：StringReplace("123412","12","00")\n【返回值】："003400"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Replace",
        "name": "字符串替换",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "oldValue",
                "name": "被替换值",
                "type": "string"
            },
            {
                "key": "newValue",
                "name": "替换值",
                "type": "string"
            }
        ],
        "description": `【功能】：将源字符串中的"被替换值"替换为"替换值"\n【original】：源字符串\n【oldValue】：被替换值\n【newValue】：替换值\n【返回值类型】：字符串(System.String)\n【示例】：Replace("123412","12","00")\n【返回值】："003400"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringToUpper",
        "name": "转大写",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：将源字符串转换为大写\n【original】：源字符串\n【返回值类型】：字符串(System.String)\n【示例】：StringToUpper("abc")\n【返回值】："ABC"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToUpper",
        "name": "转大写",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：将源字符串转换为大写\n【original】：源字符串\n【返回值类型】：字符串(System.String)\n【示例】：ToUpper("abc")\n【返回值】："ABC"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringToLower",
        "name": "转小写",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：将源字符串转换为小写\n【original】：源字符串\n【返回值类型】：字符串(System.String)\n【示例】：StringToLower("ABC")\n【返回值】："abc"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToLower",
        "name": "转小写",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：将源字符串转换为小写\n【original】：源字符串\n【返回值类型】：字符串(System.String)\n【示例】：ToLower("ABC")\n【返回值】："abc"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringIndexOf",
        "name": "获取首个匹配字符串的索引",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：在源字符串中，检索目标字符串，找到第一个匹配的字符串后，返回其索引位置(从0开始)。如果未找到目标字符串，则返回-1，如果目标字符串为空串，返回0，如果原字符串为null，则返回-1\n【original】：源字符串\n【target】：目标字符串\n【返回值类型】：数值(System.Int32)\n【示例】：StringIndexOf("12001234012","01")\n【返回值】：3\n【示例】：StringIndexOf("12002323","34")\n【返回值】：-1\n【示例】：StringIndexOf("120011","")\n【返回值】：0`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "IndexOf",
        "name": "获取首个匹配字符串的索引",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：在源字符串中，检索目标字符串，找到第一个匹配的字符串后，返回其索引位置(从0开始)。如果未找到目标字符串，则返回-1，如果目标字符串为空串，返回0，如果原字符串为null，则返回-1\n【original】：源字符串\n【target】：目标字符串\n【返回值类型】：数值(System.Int32)\n【示例】：IndexOf("12001234012","01")\n【返回值】：3\n【示例】：IndexOf("12002323","34")\n【返回值】：-1\n【示例】：IndexOf("120011","")\n【返回值】：0`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringLastIndexOf",
        "name": "获取最后一个匹配字符串的索引",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：在源字符串中，检索目标字符串，找到最后一个匹配的字符串后，返回其索引位置(从0开始)。如果未找到目标字符串，则返回-1，如果目标字符串为空串，返回0，如果原字符串为null，则返回-1\n【original】：源字符串\n【value】：目标字符串\n【返回值类型】：数值(System.Int32)\n【示例】：StringLastIndexOf("12001234012","01")\n【返回值】：8\n【示例】：StringLastIndexOf("12002323","34")\n【返回值】：-1\n【示例】：StringLastIndexOf("120011","")\n【返回值】：0`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastIndexOf",
        "name": "获取最后一个匹配字符串的索引",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            },
            {
                "key": "target",
                "name": "目标字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：在源字符串中，检索目标字符串，找到最后一个匹配的字符串后，返回其索引位置(从0开始)。如果未找到目标字符串，则返回-1，如果目标字符串为空串，返回0，如果原字符串为null，则返回-1\n【original】：源字符串\n【value】：目标字符串\n【返回值类型】：数值(System.Int32)\n【示例】：LastIndexOf("12001234012","01")\n【返回值】：8\n【示例】：LastIndexOf("12002323","34")\n【返回值】：-1\n【示例】：LastIndexOf("120011","")\n【返回值】：0`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringTrimStart",
        "name": "删除头部空格",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：删除源字符串的头部空格\n【original】：源字符串\n【返回值类型】：字符串(System.String)\n【示例】：StringTrimStart("    ABC")\n【返回值】："ABC"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "TrimStart",
        "name": "删除头部空格",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：删除源字符串的头部空格\n【original】：源字符串\n【返回值类型】：字符串(System.String)\n【示例】：TrimStart("    ABC")\n【返回值】："ABC"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringTrimEnd",
        "name": "删除尾部空格",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：删除源字符串的尾部空格\n【original】：源字符串\n【返回值类型】：字符串(System.String)\n【示例】：StringTrimEnd("ABC    ")\n【返回值】："ABC"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "TrimEnd",
        "name": "删除尾部空格",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：删除源字符串的尾部空格\n【original】：源字符串\n【返回值类型】：字符串(System.String)\n【示例】：TrimEnd("ABC    ")\n【返回值】："ABC"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringTrim",
        "name": "删除头尾部空格",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：删除源字符串的头尾部空格\n【original】：源字符串\n【返回值类型】：字符串(System.String)\n【示例】：StringTrim("    ABC    ")\n【返回值】："ABC"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Trim",
        "name": "删除头尾部空格",
        "params": [
            {
                "key": "original",
                "name": "源字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：删除源字符串的头尾部空格\n【original】：源字符串\n【返回值类型】：字符串(System.String)\n【示例】：Trim("    ABC    ")\n【返回值】："ABC"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "CreateGuid",
        "name": "唯一标识符",
        "params": [],
        "description": `【功能】：获取唯一标识符Guid\n【返回值类型】：字符串(System.String)\n【示例】：CreateGuid()\n【返回值】："7722A793-14DF-49CD-B7E7-DF4987YYD2"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "StringAdd",
        "name": "字符串相加",
        "params": [
            {
                "key": "firstStr",
                "name": "第一个字符串",
                "type": "string"
            },
            {
                "key": "secondStr",
                "name": "第二个字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：将若干字符串拼接成一个新字符串，最多支持10个字符串拼接，最少2个；编辑器中默认插入两个，多个字符串可手动添加\n【返回值类型】：字符串(System.String)\n【示例】：StringAdd("Hello","World")\n【返回值】："HelloWord"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Add",
        "name": "字符串相加",
        "params": [
            {
                "key": "firstStr",
                "name": "第一个字符串",
                "type": "string"
            },
            {
                "key": "secondStr",
                "name": "第二个字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：将若干字符串拼接成一个新字符串，最多支持10个字符串拼接，最少2个；编辑器中默认插入两个，多个字符串可手动添加\n【返回值类型】：字符串(System.String)\n【示例】：Add("Hello","World")\n【返回值】："HelloWord"`,
        "type": "String",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "SumByProp",
        "name": "按属性求和",
        "params": [
            {
                "key": "chainedPropName",
                "name": "实体集合属性名",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "类型属性名",
                "type": "string"
            }
        ],
        "description": `【功能】：统计某一属性的和\n【chainedPropName】：对应实体集合的属性名,链式形式组织在一起的属性名,例如:Order.OrderItems\n【propertyName】：类型属性名;\n【返回值类型】：数值(System.Double)\n【示例】：SumByProp("Order.OrderItems","Quantity")"`,
        "type": "Aggregate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "CountByProp",
        "name": "列行数",
        "params": [
            {
                "key": "chainedPropName",
                "name": "实体集合属性名",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "类型属性名",
                "type": "string"
            }
        ],
        "description": `【功能】：统计指定字段所在表的记录数\n【chainedPropName】：对应实体集合的属性名,链式形式组织在一起的属性名,例如：Order.OrderItems\n【propertyName】：类型属性名\n【返回值类型】：数值(System.Int32);\n【示例】：CountByProp("Order.OrderItems","Quantity")`,
        "type": "Aggregate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "AvgByProp",
        "name": "列平均值",
        "params": [
            {
                "key": "chainedPropName",
                "name": "实体集合属性名",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "类型属性名",
                "type": "string"
            }
        ],
        "description": `【功能】：统计某一数值列的平均值\n【chainedPropName】：对应实体集合的属性名,链式形式组织在一起的属性名,例如:Order.OrderItems\n【propertyName】：类型属性名\n【返回值类型】：数值(System.Double)\n【示例】：AvgByProp("Order.OrderItems","Quantity")`,
        "type": "Aggregate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "MaxByProp",
        "name": "列最大值",
        "params": [
            {
                "key": "chainedPropName",
                "name": "实体集合属性名",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "类型属性名",
                "type": "string"
            }
        ],
        "description": `【功能】：统计某一数值列的最大值\n【chainedPropName】：对应实体集合的属性名,链式形式组织在一起的属性名,例如:Order.OrderItems\n【propertyName】：类型属性名\n【返回值类型】：数值(System.Double)\n【示例】：MaxByProp("Order.OrderItems","Quantity")`,
        "type": "Aggregate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "MinByProp",
        "name": "列最小值",
        "params": [
            {
                "key": "chainedPropName",
                "name": "实体集合属性名",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "类型属性名",
                "type": "string"
            }
        ],
        "description": `【功能】：统计某一数值列的最小值\n【chainedPropName】：对应实体集合的属性名,链式形式组织在一起的属性名,例如:Order.OrderItems\n【propertyName】：类型属性名\n【返回值类型】：数值(System.Double)\n【示例】：MinByProp("Order.OrderItems","Quantity")`,
        "type": "Aggregate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "IsNull",
        "name": "是空值",
        "params": [
            {
                "key": "obj",
                "name": "检测对象",
                "type": "object"
            }
        ],
        "description": `【功能】：判断指定的对象是否为空值(Null)\n【obj】：要检测的对象\n【返回值类型】：布尔值(System.Boolean)\n【示例】：IsNull("语义化表达式")\n【返回值】：False`,
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "IsNullOrWhiteSpace",
        "name": "是空串",
        "params": [
            {
                "key": "obj",
                "name": "待判断字符串",
                "type": "string"
            }
        ],
        "description": `【功能】：判断指定的字符串是否为null、空还是仅有空格组成\n【obj】：要判断的字符串\n【返回值类型】：布尔值(System.Boolean)\n【示例】：IsNullOrWhiteSpace("语义化表达式")\n【返回值】：False\n【示例】：IsNullOrWhiteSpace(null)\n【返回值】：True\n【示例】：\nIsNullOrWhiteSpace(" ")【返回值】：True`,
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "IsNaN",
        "name": "非数字",
        "params": [
            {
                "key": "obj",
                "name": "检测对象",
                "type": "object"
            }
        ],
        "description": `【功能】：判断指定的对象是否可以转换成数字(System.Double)\n【obj】：要检测的对象\n【返回值类型】：布尔值(System.Boolean)\n【示例】：IsNaN("语义化表达式")【返回值】：True\n【示例】：IsNaN(12.3)【返回值】：False\n【示例】：\nIsNaN("123")【返回值】：False`,
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "IsNumber",
        "name": "是数字",
        "params": [
            {
                "key": "obj",
                "name": "检测对象",
                "type": "object"
            }
        ],
        "description": `【功能】：判断指定的对象是否可以转换成数字(System.Double)\n【obj】：要检测的对象\n【返回值类型】：布尔值(System.Boolean)\n【示例】：IsNumber(\"语义化表达式\")【返回值】：False\n【示例】：IsNumber(12.3)【返回值】：True\n【示例】：\nIsNumber("123")【返回值】：True`,
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "IsExistRecord",
        "name": "有无等于匹配字符",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            },
            {
                "key": "match",
                "name": "匹配值",
                "type": "string"
            }
        ],
        "description": `【功能】：根据链式名称判断该表的指定字段集合中是否存在等于匹配字符的值\n【chainedPropName】链式名称\n【propertyName】属性名\n【match】匹配值\n【返回值类型】：对象(System.Boolean)\n【示例】：IsExistRecord("salesorder.items","code","lc")\n【返回值】：False/True`,
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ListGreaterThan",
        "name": "有无大于匹配字符",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            },
            {
                "key": "match",
                "name": "匹配值",
                "type": "double"
            }
        ],
        "description": `【功能】：根据链式名称判断该表的指定字段集合中是否存在大于匹配字符的值\n【chainedPropName】链式名称\n【propertyName】属性名\n【match】匹配值\n【返回值类型】：对象(System.Boolean)\n【示例】：ListGreaterThan("salesorder.items","num",100)\n【返回值】：False/True`,
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ListLessThan",
        "name": "有无小于匹配字符",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            },
            {
                "key": "match",
                "name": "匹配值",
                "type": "double"
            }
        ],
        "description": `【功能】：根据链式名称判断该表的指定字段集合中是否存在小于匹配字符的值\n【chainedPropName】链式名称\n【propertyName】属性名\n【match】匹配值\n【返回值类型】：对象(System.Boolean)\n【示例】：ListLessThan("salesorder.items","num",100)\n【返回值】：False/True`,
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ListStartWith",
        "name": "有无开头是匹配字符",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            },
            {
                "key": "match",
                "name": "匹配值",
                "type": "string"
            }
        ],
        "description": `【功能】：根据链式名称判断该表的指定字段集合中是否存在开头是匹配字符的值\n【chainedPropName】链式名称\n【propertyName】属性名\n【match】匹配值\n【返回值类型】：对象(System.Boolean)\n【示例】：ListStartWith("salesorder.items","code","lc")\n【返回值】：False/True`,
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ListEndWith",
        "name": "有无结尾是匹配字符",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            },
            {
                "key": "match",
                "name": "匹配值",
                "type": "string"
            }
        ],
        "description": `【功能】：根据链式名称判断该表的指定字段集合中是否存在结尾是匹配字符的值\n【chainedPropName】链式名称\n【propertyName】属性名\n【match】匹配值\n【返回值类型】：对象(System.Boolean)\n【示例】：ListEndWith("salesorder.items","code","lc")\n【返回值】：False/True`,
        "type": "Estimate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "DateTimeAddDays",
        "name": "增加指定天数",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            },
            {
                "key": "value",
                "name": "增加天数",
                "type": "int"
            }
        ],
        "description": `【功能】：在指定的日期时间上增加指定的天数\n【date】：日期时间\n【value】：增加天数\n【返回值类型】：日期时间(System.DateTime)\n【示例】：DateTimeAddDays("2019-01-30T10:00:00+08:00",2)\n【返回值】:2019-02-01T10:00+08:00`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "DateTimeAddYears",
        "name": "增加指定年数",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            },
            {
                "key": "value",
                "name": "增加年数",
                "type": "int"
            }
        ],
        "description": `【功能】：在指定的日期时间上增加指定的年数\n【date】：日期时间\n【value】：增加年数\n【返回值类型】：日期时间(System.DateTime)\n【示例】：DateTimeAddYears("2019-01-30T10:00:00+08:00",2)\n【返回值】:2021-01-30T10:00+08:00`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "DateTimeAddMonths",
        "name": "增加指定月数",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            },
            {
                "key": "value",
                "name": "增加月数",
                "type": "int"
            }
        ],
        "description": `【功能】：在指定的日期时间上增加指定的月数\n【date】：日期时间\n【value】：增加月数\n【返回值类型】：日期时间(System.DateTime)\n【示例】：DateTimeAddMonths("2019-01-30T10:00:00+08:00",1)\n【返回值】:2019-02-28T10:00+08:00`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetFirstDayOfMonth",
        "name": "获取指定日期时间所在月第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：获取指定日期时间值所在月的第一天的日期\n【date】：日期时间\n【返回值类型】：日期时间(System.DateTime)\n【示例】：GetFirstDayOfMonth("2019-01-30T10:00:00+08:00")\n【返回值】:2019-01-01T10:00+08:00`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetLastDayOfMonth",
        "name": "获取指定日期时间所在月最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：获取指定日期时间值所在月的最后一天的日期\n【date】：日期时间\n【返回值类型】：日期时间(System.DateTime)\n【示例】：GetLastDayOfMonth("2019-01-30T10:00:00+08:00")\n【返回值】:2019-01-31T10:00+08:00`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetDateTimeNow",
        "name": "获取当前日期时间",
        "params": [],
        "description": `【功能】：获取当前的日期时间\n【返回值类型】：日期时间(System.DateTime)\n【示例】：GetDateTimeNow()\n【返回值】:2019-01-30T12:00:00.000+08:00`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "CompareDate",
        "name": "比较两个时间",
        "params": [
            {
                "key": "str1",
                "name": "日期时间1",
                "type": "string"
            },
            {
                "key": "str2",
                "name": "日期时间2",
                "type": "string"
            }
        ],
        "description": `【功能】：比较两个时间值，并返回一个值，此值表示第一个时间是早于、等于还是晚于第二个时间的整数\n【str1】：日期时间1\n【str2】：日期时间2\n【返回值类型】：数值(System.int32)\n\t日期1早于日期2，返回-1\n\t日期1等于日期2，返回0\n\t日期1晚于日期2，返回1\n【示例】：CompareDate("2019-01-30T10:00:00+08:00","2019-02-01T10:00:00+08:00")\n【返回值】：-1`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FormatDate",
        "name": "格式化当前日期",
        "params": [
            {
                "key": "format",
                "name": "格式串",
                "type": "string"
            }
        ],
        "description": `【功能】：使用指定的格式将当前日期时间值转换为其等效的字符串表示形式\n【format】：格式串\n【返回值类型】：字符串(System.String)\n【示例】：FormatDate("yyyy-MM-dd HH:mm:ss")\n【返回值】："2019-01-30 01:00:00"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FormatDefineDate",
        "name": "格式化指定日期",
        "params": [
            {
                "key": "format",
                "name": "格式串",
                "type": "string"
            },
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：使用指定的格式将指定的日期时间值转换为其等效的字符串表示形式\n【format】：格式串\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：FormatDefineDate("yyyy-MM-dd HH:mm:ss","2019-01-30 01:00:00")\n【返回值】："2019-01-30 01:00:00"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Today",
        "name": "今天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的今天，格式为"yyyy-MM-dd"形式，如果不传递参数，则默认返回当前系统日期的今天\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：Today("2019-01-30T10:00:00+08:00")\n【返回值】："2019-01-30"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Yesterday",
        "name": "昨天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的昨天，格式为"yyyy-MM-dd"形式，如果不传递参数，则默认返回当前系统日期的昨天\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：Yesterday("2019-01-30T10:00:00+08:00")\n【返回值】："2019-01-29"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "Tomorrow",
        "name": "明天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的明天，格式为"yyyy-MM-dd"形式，如果不传递参数，则默认返回当前系统日期的明天\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：Tomorrow("2019-01-30T10:00:00+08:00")\n【返回值】："2019-01-31"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ThisYear",
        "name": "本年",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的本年，格式为"yyyy"形式，如果不传递参数，则默认返回当前系统日期的本年\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：ThisYear("2019-01-30T10:00:00+08:00")\n【返回值】："2019"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastYear",
        "name": "去年",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的去年，格式为"yyyy"形式，如果不传递参数，则默认返回当前系统日期的去年\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：LastYear("2019-01-30T10:00:00+08:00")\n【返回值】："2018"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "NextYear",
        "name": "明年",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的明年，格式为"yyyy"形式，如果不传递参数，则默认返回当前系统日期的明年\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：NextYear("2019-01-30T01:00:00+08:00")\n【返回值】："2020"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ThisMonth",
        "name": "本月",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的本月，格式为"MM"形式，如果不传递参数，则默认返回当前系统日期的本月\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：ThisMonth("2019-01-30T01:00:00+08:00")\n【返回值】："01"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastMonth",
        "name": "上月",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的上月，格式为"MM"形式，如果不传递参数，则默认返回当前系统日期的上月\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：LastMonth("2019-02-30 01:00:00")\n【返回值】："01"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "NextMonth",
        "name": "下月",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的下月，格式为"MM"形式，如果不传递参数，则默认返回当前系统日期的下月\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：NextMonth("2019-01-30T01:00:00+08:00")\n【返回值】："02"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ThisWeek",
        "name": "本周",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期在本年的第几周，格式为"ww"形式，如果不传递参数，则默认返回当前系统日期在本年的第几周，注意如果12月底的日期与下一年的1月1日在同一星期，则返回第一周\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：ThisWeek("2019-01-30T01:00:00+08:00")\n【返回值】："5"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastWeek",
        "name": "上周",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的上周，格式为"ww"形式，如果不传递参数，则默认返回当前系统日期的上周，注意如果指定日期在今年第1周，则上周为去年的最后一周\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：LastWeek("2019-01-30T01:00:00+08:00")\n【返回值】："4"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "NextWeek",
        "name": "下周",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的下周，格式为"ww"形式，如果不传递参数，则默认返回当前系统日期的下周，注意如果指定日期在今年最后一周，则下周为1\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：NextWeek("2019-01-30T01:00:00+08:00")\n【返回值】："6"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "DayOfWeek",
        "name": "星期几",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期是本周的星期几，格式为"n"，如果不传递参数，则默认返回当前系统日期是本周的星期几，注意不是返回英文缩写，而是返回数字形式。按照中国习惯，星期一为每周首日，返回值为1，星期日为每周尾日，返回值为7\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：DayOfWeek("2019-01-30T01:00:00+08:00")\n【返回值】："3"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfWeek",
        "name": "本周第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的本周第一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的本周第一天。按照中国习惯，星期一为每周首日，星期日为每周尾日\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：FirstDayOfWeek("2019-01-30T01:00:00+08:00")\n【返回值】："2019-01-28"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfWeek",
        "name": "本周最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的本周最后一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的本周第一天。按照中国习惯，星期一为每周首日，星期日为每周尾日\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：FirstDayOfWeek("2019-01-30T01:00:00+08:00")\n【返回值】："2019-02-03"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfLastWeek",
        "name": "上周第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的上周第一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的上周第一天。按照中国习惯，星期一为每周首日，星期日为每周尾日\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：FirstDayOfLastWeek("2019-01-30T01:00:00+08:00")\n【返回值】："2019-01-21"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfLastWeek",
        "name": "上周最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的上周最后一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的上周最后一天。按照中国习惯，星期一为每周首日，星期日为每周尾日\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：LastDayOfLastWeek("2019-01-30T01:00:00+08:00")\n【返回值】："2019-01-27"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfNextWeek",
        "name": "下周第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的下周第一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的下周第一天。按照中国习惯，星期一为每周首日，星期日为每周尾日\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：FirstDayOfNextWeek("2019-01-30T01:00:00+08:00")\n【返回值】："2019-02-04"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfNextWeek",
        "name": "下周最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的下周最后一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的下周最后一天。按照中国习惯，星期一为每周首日，星期日为每周尾日\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：LastDayOfNextWeek("2019-01-30T01:00:00+08:00")\n【返回值】："2019-02-10"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfMonth",
        "name": "本月第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的本月第一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的本月第一天。\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：FirstDayOfMonth("2019-01-30T01:00:00+08:00")\n【返回值】："2019-01-01"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfMonth",
        "name": "本月最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的本月最后一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的本月最后一天。\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：LastDayOfMonth("2019-01-30T01:00:00+08:00")\n【返回值】："2019-01-31"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfLastMonth",
        "name": "上月第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的上月第一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的上月第一天。\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：FirstDayOfLastMonth("2019-01-30T01:00:00+08:00")\n【返回值】："2018-12-01"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfLastMonth",
        "name": "上月最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的上月最后一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的上月最后一天。\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：LastDayOfLastMonth("2019-01-30T01:00:00+08:00")\n【返回值】："2018-12-31"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfNextMonth",
        "name": "下月第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的下月第一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的下月第一天。\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：FirstDayOfNextMonth("2019-01-30T01:00:00+08:00")\n【返回值】："2019-02-01"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfNextMonth",
        "name": "下月最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的下月最后一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的下月最后一天。\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：LastDayOfNextMonth("2019-01-30T01:00:00+08:00")\n【返回值】："2019-02-28"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfYear",
        "name": "本年第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的本年第一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的本年第一天。\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：FirstDayOfYear("2019-01-30T01:00:00+08:00")\n【返回值】："2019-01-01"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfYear",
        "name": "本年最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的本年最后一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的本年最后一天。\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：LastDayOfYear("2019-01-30T01:00:00+08:00")\n【返回值】："2019-12-31"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfLastYear",
        "name": "上年第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的上年第一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的上年第一天。\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：FirstDayOfLastYear("2019-01-30T01:00:00+08:00")\n【返回值】："2018-01-01"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfLastYear",
        "name": "上年最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的上年最后一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的上年最后一天。\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：LastDayOfLastYear("2019-01-30T01:00:00+08:00")\n【返回值】："2018-12-31"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "FirstDayOfNextYear",
        "name": "下年第一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的下年第一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的下年第一天。\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：FirstDayOfNextYear("2019-01-30T01:00:00+08:00")\n【返回值】："2020-01-01"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "LastDayOfNextYear",
        "name": "下年最后一天",
        "params": [
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定日期的下年最后一天，格式为"yyyy-MM-dd"，如果不传递参数，则默认返回当前系统日期的下年最后一天。\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n【示例】：LastDayOfNextYear("2019-01-30T01:00:00+08:00")\n【返回值】："2020-12-31"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": "DefaultFunction"
    },
    {
        "key": "GetDate",
        "name": "返回指定形式的日期",
        "params": [
            {
                "key": "format",
                "name": "格式串",
                "type": "string"
            },
            {
                "key": "date",
                "name": "日期时间",
                "type": "string"
            }
        ],
        "description": `【功能】：返回指定形式的日期，如果不传递参数，则默认返回当前系统日期的指定形式日期。\n【format】：格式化串\n【date】：日期时间\n【返回值类型】：字符串(System.String)\n\tDD：今天\tLD：昨天\tND：明天\n\tYY：本年\tLY：上年\tNY：下年\n\tMM：本月\tLM：上月\tNM：下月\n\tWD：星期几\n\tWW：本周\tLW：上周\tNW：下周\n\tFDW：本周首日\tLDW：本周尾日\n\tFDLW：上周首日\tLDLW：上周尾日\n\tFDNW：下周首日\tLDNW：下周尾日\n\tFDM：本月首日\tLDM：本月尾日\n\tFDLM：上月首日\tLDLM：上月尾日\n\tFDNM：下月首日\tLDNM：下月尾日\n\tFDY：本年首日\tLDY：本年尾日\n\tFDLY：上年首日\tLDLY：上年尾日\n\tFDNY：下年首日\tLDNY：下年尾日\n【示例】：GetDate("LD","2019-01-30T01:00:00+08:00")\n【返回值】："2019-01-29"`,
        "type": "DateTime",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToStringX",
        "name": "转字符串",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": `【功能】：将指定的对象转换成字符串\n【obj】：待转换对象\n【返回值类型】：字符串(System.String)\n【示例】：ToStringX(123)\n【返回值】："123"`,
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToDateTime",
        "name": "转日期时间",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": `【功能】：将指定的对象转换成日期时间\n【obj】：待转换对象\n【返回值类型】：日期时间(System.OffsetDateTime)\n【示例】：ToDateTime("2019-01-30T10:00:00+08:00")\n【返回值】: 2019-01-30T10:00+08:00`,
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToBoolean",
        "name": "转布尔值",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": `【功能】：将指定对象值转换成等效的布尔值\n【obj】：待转换对象\n【返回值类型】：布尔值(System.Boolean)\n【示例】：ToBoolean("True")\n【返回值】：true`,
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToDecimal",
        "name": "转数值",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": `【功能】：将指定对象转换成十进制数值\n【obj】：待转换对象\n【返回值类型】：十进制数值(System.BigDecimal)\n【示例】：ToDecimal("40.23")\n【返回值】：40.23`,
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToBigDecimal",
        "name": "转数值",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": `【功能】：将指定对象转换成十进制数值\n【obj】：待转换对象\n【返回值类型】：十进制数值(System.BigDecimal)\n【示例】：ToBigDecimal("40.23")\n【返回值】：40.23`,
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToDouble",
        "name": "转双精度浮点数",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": `【功能】：将指定对象转换成双精度浮点数\n【obj】：待转换对象\n【返回值类型】：双精度浮点数(System.Double)\n【示例】：ToDouble("40.23")\n【返回值】：40.23`,
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToSingle",
        "name": "转单精度浮点数",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": `【功能】：将指定对象转换成单精度浮点数\n【obj】：待转换对象\n【返回值类型】：单精度浮点数(System.Float)\n【示例】：ToSingle("40.23")\n【返回值】：40.23`,
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToFloat",
        "name": "转单精度浮点数",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": `【功能】：将指定对象转换成单精度浮点数\n【obj】：待转换对象\n【返回值类型】：单精度浮点数(System.Float)\n【示例】：ToFloat("40.23")\n【返回值】：40.23`,
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToByte",
        "name": "转8位无符号整数",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": `【功能】：将指定对象转换成8位无符号整数\n【obj】：待转换对象\n【返回值类型】：8位无符号整数(System.Byte)\n【示例】：ToByte("10")\n【返回值】：10`,
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToInt16",
        "name": "转16位有符号整数",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": `【功能】：将指定对象转换成16位有符号整数\n【obj】：待转换对象\n【返回值类型】：16位有符号整数(System.Int16)\n【示例】：ToInt16("10")\n【返回值】：10`,
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToInt32",
        "name": "转32位有符号整数",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": `【功能】：将指定对象转换成32位有符号整数\n【obj】：待转换对象\n【返回值类型】：32位有符号整数(System.Int32)\n【示例】：ToInt32("10")\n【返回值】：10`,
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "ToChineseMoney",
        "name": "转中文大写金额",
        "params": [
            {
                "key": "obj",
                "name": "转换对象",
                "type": "object"
            }
        ],
        "description": `【功能】：将指定对象转换成中文大写金额\n【obj】：待转换对象\n【返回值类型】：字符串(System.String)\n【示例】：ToChineseMoney(40.23)\n【返回值】："肆拾元零贰角叁分"`,
        "type": "Translate",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetSessionValue",
        "name": "获取会话变量的值",
        "params": [
            {
                "key": "name",
                "name": "会话变量名",
                "type": "string"
            }
        ],
        "description": `【功能】：获取用户会话变量的值\n【name】：会话变量名称\n【返回值类型】：对象(System.Object)`,
        "type": "Analytical",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetContextParameter",
        "name": "获取上下文变量的值",
        "params": [
            {
                "key": "name",
                "name": "上下文变量名",
                "type": "string"
            }
        ],
        "description": `【功能】：获取上下文变量的值\n【name】：上下文变量名称\n【返回值类型】：对象(System.Object)`,
        "type": "Analytical",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "getComputeJsonData",
        "name": "根据属性名获取JSON串",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            }
        ],
        "description": `【功能】：根据链式属性名得到JSON串\n【chainedPropName】链式名称\n【propertyName】属性名\n【返回值类型】：对象(System.String)`,
        "type": "Analytical",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetComputeJsonData",
        "name": "根据属性名获取JSON串",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名称",
                "type": "string"
            },
            {
                "key": "propertyName",
                "name": "属性名",
                "type": "string"
            }
        ],
        "description": `【功能】：根据链式属性名得到JSON串\n【chainedPropName】链式名称\n【propertyName】属性名\n【返回值类型】：对象(System.String)`,
        "type": "Analytical",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetInjectedEntity",
        "name": "根据注入名获获取注入对象",
        "params": [
            {
                "key": "name",
                "name": "注入名",
                "type": "string"
            }
        ],
        "description": `【功能】：获取运行时注入的对象\n【name】注入名\n【返回值类型】：对象(System.Object)`,
        "type": "Analytical",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "GetChainedPropertyValue",
        "name": "根据属性名获取属性值",
        "params": [
            {
                "key": "chainedPropName",
                "name": "链式名",
                "type": "string"
            }
        ],
        "description": `【功能】：根据链式属性名获取属性值\n【chainedPropName】链式名\n【返回值类型】：对象(System.Object)`,
        "type": "Analytical",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "abs",
        "name": "绝对值",
        "params": [
            {
                "key": "value",
                "name": "数值",
                "type": "double"
            }
        ],
        "description": `【功能】：返回一个数值的绝对值\n【value】：计算数值\n【返回值类型】：数值(System.Double)\n【示例】：abs(-1.5)\n【返回值】：1.5`,
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "Round",
        "name": "四舍五入",
        "params": [
            {
                "key": "value1",
                "name": "数值",
                "type": "double"
            },
            {
                "key": "value2",
                "name": "小数位数",
                "type": "int32"
            }
        ],
        "description": `【功能】：把数四舍五入为指定精度的数值\n【value1】：数值 \n【value2】：四舍五入小数位数\n【返回值类型】：数值(System.Double)\n【示例】：\nRound(1.52,1)\n【返回值】：1.5`,
        "type": "Math",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "round",
        "name": "四舍五入",
        "params": [
            {
                "key": "value1",
                "name": "数值",
                "type": "double"
            },
            {
                "key": "value2",
                "name": "小数位数",
                "type": "int32"
            }
        ],
        "description": `【功能】：把数四舍五入为指定精度的数值\n【value1】：数值 \n【value2】：四舍五入小数位数\n【返回值类型】：数值(System.Double)\n【示例】：round(1.52,1)\n【返回值】：1.5`,
        "type": "Math",
        "keyFront": "DefaultFunction",
        "className": ""
    },
    {
        "key": "random",
        "name": "随机数",
        "params": [],
        "description": `【功能】：随机生成大于等于 0.0 且小于 1.0 的伪随机值\n【返回值类型】：数值(System.Double)\n【示例】：random()\n【返回值】：0.624508349393`,
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "sin",
        "name": "正弦",
        "params": [
            {
                "key": "value",
                "name": "数值",
                "type": "double"
            }
        ],
        "description": `【功能】：返回一个数字的正弦值\n【value】：计算数值\n【返回值类型】：数值(System.Double)\n一个以弧度表示的角\n【示例】：sin(1)\n【返回值】：0.84187098`,
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "cos",
        "name": "余弦",
        "params": [
            {
                "key": "value",
                "name": "数值",
                "type": "double"
            }
        ],
        "description": `【功能】：返回一个数字的余弦值\n【value】：计算数值\n【返回值类型】：数值(System.Double)\n一个以弧度表示的角\n【示例】：cos(3)\n【返回值】：-0.9899924`,
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "ceil",
        "name": "向上舍入",
        "params": [
            {
                "key": "value",
                "name": "数值",
                "type": "double"
            }
        ],
        "description": `【功能】：返回大于或等于函数参数，并且与之最接近的整数\n【value】：计算数值\n【返回值类型】：数值(System.Int32)\n【示例】：ceil(3.2)\n【返回值】：4`,
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "floor",
        "name": "向下舍入",
        "params": [
            {
                "key": "value",
                "name": "数值",
                "type": "double"
            }
        ],
        "description": `【功能】：返回小于或等于函数参数，并且与之最接近的整数\n【value】：计算数值\n【返回值类型】：数值(System.Int32)\n【示例】：floor(3.2)\n【返回值】：3`,
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "BankerRound",
        "name": "银行家舍入",
        "params": [
            {
                "key": "value1",
                "name": "数值",
                "type": "double"
            },
            {
                "key": "value2",
                "name": "小数位数",
                "type": "int32"
            }
        ],
        "description": `【功能】：使用银行家舍入标准，把指定数字舍入为最接近的数字\n【value1】：要舍入的数值\n【value2】：要舍入的小数位数\n【返回值类型】：数值(System.Double)\n【示例】：BankerRound(36.124,2)\n【返回值】：36.12`,
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    },
    {
        "key": "bankerRound",
        "name": "银行家舍入",
        "params": [
            {
                "key": "value1",
                "name": "数值",
                "type": "double"
            },
            {
                "key": "value2",
                "name": "小数位数",
                "type": "int32"
            }
        ],
        "description": `【功能】：使用银行家舍入标准，把指定数字舍入为最接近的数字\n【value1】：要舍入的数值\n【value2】：要舍入的小数位数\n【返回值类型】：数值(System.Double)\n【示例】：bankerRound(36.124,2)\n【返回值】：36.12`,
        "type": "Math",
        "keyFront": "Math",
        "className": ""
    }
];
