// @gsp-cmp/ubcc-param-data-ui/esm2015/lib/controls/param-control/checkbox-control/checkbox-control.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/controls/param-control/checkbox-control/checkbox-control.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { UbccParamDataUtils } from '../../../utils/param-data-utils';
export class CheckboxControlComponent {
    /**
     * @param {?} service
     */
    constructor(service) {
        this.service = service;
        this.canEdit = true;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.param = this.param || {};
        this.checkboxId = this.service.newGuid();
    }
    /**
     * @return {?}
     */
    getValue() {
        if (!this.param || !this.param.value) {
            return false;
        }
        /** @type {?} */
        let value = String(this.param.value).toLowerCase();
        return value === 'true' ? true : false;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    enable(value) {
        this.param.value = value;
    }
    /**
     * @return {?}
     */
    isReadonly() {
        if (!this.canEdit) {
            return true;
        }
        else {
            return !this.param || !this.param.readonly ? false : true;
        }
    }
    /**
     * @return {?}
     */
    show() {
        return !this.param || !this.param.description ? false : true;
    }
}
CheckboxControlComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-checkbox-control',
                template: "<div class=\"farris-form-controls-inline farris-group-wrap col-12\">\n  <div class=\"form-group farris-form-group\">\n    <label class=\"col-form-label col-4\">\n      <span class=\"farris-label-text\" style=\"float:right\">{{param.name}}</span>\n    </label>\n    <div class=\"farris-input-wrap col-6\">\n      <div class=\"custom-control custom-checkbox\">\n        <input [checked]=\"getValue()\" class=\"custom-control-input\" id=\"{{checkboxId}}Checkbox\" \n          type=\"checkbox\" (click)=\"enable($event.target.checked)\" [disabled]=\"isReadonly()\">\n        <label class=\"custom-control-label\" for=\"{{checkboxId}}Checkbox\"></label>\n      </div>\n    </div>\n    <span *ngIf=\"show()\" farrisTooltip [placement]=\"'right-top'\" [content]=\"param.description\" class=\"f-icon f-icon-info\"></span>\n  </div>\n</div>",
                styles: [""]
            }] }
];
/** @nocollapse */
CheckboxControlComponent.ctorParameters = () => [
    { type: UbccParamDataUtils }
];
CheckboxControlComponent.propDecorators = {
    param: [{ type: Input }],
    canEdit: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    CheckboxControlComponent.prototype.param;
    /** @type {?} */
    CheckboxControlComponent.prototype.canEdit;
    /** @type {?} */
    CheckboxControlComponent.prototype.checkboxId;
    /**
     * @type {?}
     * @private
     */
    CheckboxControlComponent.prototype.service;
}
