// @farris/ui-time-picker/esm5/lib/services/time-picker.locale.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { LocaleService } from '@farris/ui-locale';
import * as i0 from "@angular/core";
import * as i1 from "@farris/ui-locale";
var TimePickerLocaleService = /** @class */ (function () {
    function TimePickerLocaleService(localeService) {
        this.localeService = localeService;
        this.localeConfig = {
            'zh-CHS': {
                placeholder: '请选择时间',
                time: {
                    hour: '时', minute: '分', seconds: '秒'
                }
            },
            'en': {
                placeholder: 'Please select a time',
                time: {
                    hour: 'h', minute: 'min', seconds: 's'
                }
            },
            'zh-CHT': {
                placeholder: '請選擇時間',
                time: {
                    hour: '時', minute: '分', seconds: '秒'
                }
            }
        };
    }
    /**
     * @return {?}
     */
    TimePickerLocaleService.prototype.getLocaleConfig = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var locale = this.localeService.localeId;
        if (locale && this.localeConfig[locale]) {
            return this.localeConfig[locale];
        }
        else {
            return this.localeConfig['zh-CHS'];
        }
    };
    TimePickerLocaleService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TimePickerLocaleService.ctorParameters = function () { return [
        { type: LocaleService }
    ]; };
    /** @nocollapse */ TimePickerLocaleService.ngInjectableDef = i0.defineInjectable({ factory: function TimePickerLocaleService_Factory() { return new TimePickerLocaleService(i0.inject(i1.LocaleService)); }, token: TimePickerLocaleService, providedIn: "root" });
    return TimePickerLocaleService;
}());
export { TimePickerLocaleService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TimePickerLocaleService.prototype.localeConfig;
    /**
     * @type {?}
     * @private
     */
    TimePickerLocaleService.prototype.localeService;
}
