// @gsp-sys/rac-common/esm5/lib/raccommon.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { GridModule } from '@progress/kendo-angular-grid';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { LayoutModule as KendoLayoutModule } from '@progress/kendo-angular-layout';
import { TreeViewModule } from '@progress/kendo-angular-treeview';
import { ButtonsModule } from '@progress/kendo-angular-buttons';
import { ModalModule } from '@farris/ui-modal';
import { MessagerModule } from '@farris/ui-messager';
import { LayoutModule } from '@farris/ui-layout';
import { LookupModule } from '@farris/ui-lookup';
import { FarrisPanelModule } from '@farris/ui-panel';
import { TreeTableModule } from '@farris/ui-treetable';
import { DataTableModule } from '@farris/ui-datatable';
import { DatagridModule } from '@farris/ui-datagrid';
import { AuthObjService } from './common/service/authobj.service';
import { AuresultService } from './common/service/auresult-service';
import { LocalRepository } from './common/util/local-repository';
import { RacCommonUtil } from './common/util/rac-util';
// import { AuresultComponent } from './component/auresult/auresult.component';
import { DataHelpComponent } from './component/data-help/data-help.component';
// import { AuthentryGridComponent } from './component/authentry/authentry-grid/authentry-grid.component';
import { AuobjHelpComponent } from './component/auobj/auobj-help/auobj-help.component';
import { AuobjTreeComponent } from './component/auobj/auobj-tree/auobj-tree.component';
import { AuthfieldHelpComponent } from './component/authfield/authfield-help/authfield-help.component';
import { AuthfieldTreeComponent } from './component/authfield/authfield-tree/authfield-tree.component';
import { ReassignGridComponent } from './component/reassign/reassign-grid/reassign-grid.component';
import { ReassignHelpComponent } from './component/reassign/reassign-help/reassign-help.component';
import { ReAssignService } from './common/service/reassign.service';
import { AuthfieldService } from './common/service/authfield-service';
var RacCommonModule = /** @class */ (function () {
    function RacCommonModule() {
    }
    RacCommonModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DataHelpComponent,
                        AuobjHelpComponent,
                        AuobjTreeComponent,
                        AuthfieldHelpComponent,
                        AuthfieldTreeComponent,
                        ReassignGridComponent,
                        ReassignHelpComponent,
                    ],
                    imports: [
                        CommonModule,
                        HttpClientModule,
                        ButtonsModule,
                        FormsModule,
                        ReactiveFormsModule,
                        GridModule,
                        DatagridModule,
                        KendoLayoutModule,
                        LayoutModule,
                        FarrisPanelModule,
                        TreeTableModule,
                        DataTableModule,
                        TreeViewModule,
                        LookupModule,
                        ModalModule.forRoot(),
                        MessagerModule.forRoot()
                    ],
                    entryComponents: [
                        DataHelpComponent,
                        AuobjHelpComponent,
                        AuobjTreeComponent,
                        AuthfieldHelpComponent,
                        AuthfieldTreeComponent,
                        ReassignHelpComponent,
                        ReassignGridComponent
                    ],
                    providers: [
                        RacCommonUtil,
                        LocalRepository,
                        AuresultService,
                        AuthObjService,
                        ReAssignService,
                        AuthfieldService
                    ],
                    exports: [
                        DataHelpComponent,
                        AuobjHelpComponent,
                        AuobjTreeComponent,
                        AuthfieldHelpComponent,
                        AuthfieldTreeComponent,
                        ReassignHelpComponent,
                        ReassignGridComponent
                    ]
                },] }
    ];
    return RacCommonModule;
}());
export { RacCommonModule };
