// @farris/extend-fileupload-adapt-unifile/esm5/lib/ffilepreview-adapt-unifile.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ffilepreview-adapt-unifile.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input } from '@angular/core';
import { FfilepreviewAdaptUnifileService } from './ffilepreview-adapt-unifile.service';
var FFilePreviewAdaptUnifileDirective = /** @class */ (function () {
    function FFilePreviewAdaptUnifileDirective(previewSer) {
        this.previewSer = previewSer;
        /**
         * 处理扩展的服务器端配置
         */
        this._extendServeConfig = null;
    }
    /**
     * 捕获预览事件
     * @param fileInfo
     */
    /**
     * 捕获预览事件
     * @param {?} fileInfo
     * @return {?}
     */
    FFilePreviewAdaptUnifileDirective.prototype.filePreviewEventHandler = /**
     * 捕获预览事件
     * @param {?} fileInfo
     * @return {?}
     */
    function (fileInfo) {
        // 可以通过属性 extendServerConfig设置，也可以通过一体化组件的Input属性配置
        /** @type {?} */
        var previewOpts = Object.assign(this.extendServerConfig || {});
        previewOpts.options = Object.assign(previewOpts.options || {}, { showDownload: !!fileInfo.showDownload }, { showComments: !!fileInfo.showComments }, { modeless: !!fileInfo.modeless });
        if (previewOpts.options.showHeader === undefined) {
            previewOpts.options.showHeader = true;
        }
        if (previewOpts.options.showFileList === undefined) {
            previewOpts.options.showFileList = true;
        }
        if (previewOpts.options.showComments === undefined) {
            previewOpts.options.showComments = false;
        }
        if (previewOpts.options.modeless === undefined) {
            previewOpts.options.modeless = false;
        }
        // 支持修订模式
        // 文件扩展名
        /** @type {?} */
        var extendName = fileInfo.name.substr(fileInfo.name.lastIndexOf('.') + 1).toLowerCase();
        /** @type {?} */
        var enablePreviews = ['doc', 'docx', 'xls', 'xlsx', 'ppt', 'pptx', 'jpg', 'jpeg', 'png', 'gif', 'bmp', 'pdf', 'txt', 'wps', 'wpt', 'et', 'dps'];
        if (enablePreviews.includes(extendName)) {
            this.previewSer.previewFile(fileInfo, previewOpts);
        }
        else {
            /** @type {?} */
            var msg = this.previewSer.getLocalStr("previewAdapt.notSupport");
            if (this.previewSer.notifySer) {
                this.previewSer.notifySer.warning(msg);
            }
            else {
                alert(msg);
            }
        }
    };
    /**
     * 捕获下载事件
     * @param fileInfo
     */
    /**
     * 捕获下载事件
     * @param {?} downloadEv
     * @return {?}
     */
    FFilePreviewAdaptUnifileDirective.prototype.fileDownloadEventHandler = /**
     * 捕获下载事件
     * @param {?} downloadEv
     * @return {?}
     */
    function (downloadEv) {
        // 如果是批量下载, downloadEv={fileInfos:[],name:}
        if (downloadEv && downloadEv['fileInfos'].length > 0) {
            if (downloadEv['fileInfos'].length > 1) {
                this.previewSer.multiDownloadFilesWidthName(downloadEv['fileInfos'], downloadEv['name'], this.extendServerConfig);
            }
            else {
                this.previewSer.downloadFile(downloadEv['fileInfos'][0], this.extendServerConfig);
            }
        }
    };
    Object.defineProperty(FFilePreviewAdaptUnifileDirective.prototype, "extendServerConfig", {
        get: /**
         * @return {?}
         */
        function () {
            return this._extendServeConfig;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._extendServeConfig = value;
            this.previewSer.setPreviwExtendServerConfig(value);
        },
        enumerable: true,
        configurable: true
    });
    FFilePreviewAdaptUnifileDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[fFilePreviewAdaptUnifile]',
                    providers: [FfilepreviewAdaptUnifileService]
                },] }
    ];
    /** @nocollapse */
    FFilePreviewAdaptUnifileDirective.ctorParameters = function () { return [
        { type: FfilepreviewAdaptUnifileService }
    ]; };
    FFilePreviewAdaptUnifileDirective.propDecorators = {
        filePreviewEventHandler: [{ type: HostListener, args: ['filePreviewEvent', ['$event'],] }],
        fileDownloadEventHandler: [{ type: HostListener, args: ['fileDownloadEvent', ['$event'],] }],
        extendServerConfig: [{ type: Input }]
    };
    return FFilePreviewAdaptUnifileDirective;
}());
export { FFilePreviewAdaptUnifileDirective };
if (false) {
    /**
     * 处理扩展的服务器端配置
     * @type {?}
     * @private
     */
    FFilePreviewAdaptUnifileDirective.prototype._extendServeConfig;
    /**
     * @type {?}
     * @private
     */
    FFilePreviewAdaptUnifileDirective.prototype.previewSer;
}
