// @farris/extend-fileupload-adapt-unifile/esm5/lib/ffilepreview-adapt-unifile-simple.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ffilepreview-adapt-unifile-simple.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input } from '@angular/core';
import { FfilepreviewAdaptUnifileService } from './ffilepreview-adapt-unifile.service';
var FFileAdaptPreviewFileDirective = /** @class */ (function () {
    function FFileAdaptPreviewFileDirective(previewSer) {
        this.previewSer = previewSer;
        // 是否禁用预览
        this.viewDisabled = false;
        // 处理扩展的服务器端配置
        this.extendServerConfig = null;
    }
    /**
     * 捕获预览事件
     * @param fileInfo
     */
    /**
     * 捕获预览事件
     * @return {?}
     */
    FFileAdaptPreviewFileDirective.prototype.filePreviewEventHandler = /**
     * 捕获预览事件
     * @return {?}
     */
    function () {
        if (!this.fileInfo || this.viewDisabled) {
            return;
        }
        this.previewSer.previewFile(this.fileInfo, this.extendServerConfig);
    };
    FFileAdaptPreviewFileDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[fFileAdaptPreviewFile]'
                },] }
    ];
    /** @nocollapse */
    FFileAdaptPreviewFileDirective.ctorParameters = function () { return [
        { type: FfilepreviewAdaptUnifileService }
    ]; };
    FFileAdaptPreviewFileDirective.propDecorators = {
        fileInfo: [{ type: Input, args: ['fFileAdaptPreviewFile',] }],
        filePreviewEventHandler: [{ type: HostListener, args: ['click', ['$event'],] }],
        viewDisabled: [{ type: Input }],
        extendServerConfig: [{ type: Input }]
    };
    return FFileAdaptPreviewFileDirective;
}());
export { FFileAdaptPreviewFileDirective };
if (false) {
    /** @type {?} */
    FFileAdaptPreviewFileDirective.prototype.fileInfo;
    /** @type {?} */
    FFileAdaptPreviewFileDirective.prototype.viewDisabled;
    /** @type {?} */
    FFileAdaptPreviewFileDirective.prototype.extendServerConfig;
    /**
     * @type {?}
     * @private
     */
    FFileAdaptPreviewFileDirective.prototype.previewSer;
}
var FFileAdaptDownloadFileDirective = /** @class */ (function () {
    function FFileAdaptDownloadFileDirective(previewSer) {
        this.previewSer = previewSer;
        //重命名
        this.zipName = '';
        // 是否禁用下载
        this.downloadDisabled = false;
        // 处理扩展的服务器端配置
        this.extendServerConfig = null;
        // 是否启用批量下载
        this.enableMulti = false;
    }
    /**
     * 捕获预览事件
     * @param fileInfo
     */
    /**
     * 捕获预览事件
     * @return {?}
     */
    FFileAdaptDownloadFileDirective.prototype.filePreviewEventHandler = /**
     * 捕获预览事件
     * @return {?}
     */
    function () {
        if (!this.fileInfo || this.downloadDisabled) {
            return;
        }
        if (this.enableMulti && this.fileInfo instanceof Array) {
            this.previewSer.multiDownloadFilesWidthName(this.fileInfo, this.zipName, this.extendServerConfig);
        }
        else {
            this.previewSer.downloadFile(this.fileInfo, this.extendServerConfig);
        }
    };
    FFileAdaptDownloadFileDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[fFileAdaptDownloadFile]'
                },] }
    ];
    /** @nocollapse */
    FFileAdaptDownloadFileDirective.ctorParameters = function () { return [
        { type: FfilepreviewAdaptUnifileService }
    ]; };
    FFileAdaptDownloadFileDirective.propDecorators = {
        fileInfo: [{ type: Input, args: ['fFileAdaptDownloadFile',] }],
        filePreviewEventHandler: [{ type: HostListener, args: ['click', ['$event'],] }],
        zipName: [{ type: Input }],
        downloadDisabled: [{ type: Input }],
        extendServerConfig: [{ type: Input }],
        enableMulti: [{ type: Input }]
    };
    return FFileAdaptDownloadFileDirective;
}());
export { FFileAdaptDownloadFileDirective };
if (false) {
    /** @type {?} */
    FFileAdaptDownloadFileDirective.prototype.fileInfo;
    /** @type {?} */
    FFileAdaptDownloadFileDirective.prototype.zipName;
    /** @type {?} */
    FFileAdaptDownloadFileDirective.prototype.downloadDisabled;
    /** @type {?} */
    FFileAdaptDownloadFileDirective.prototype.extendServerConfig;
    /** @type {?} */
    FFileAdaptDownloadFileDirective.prototype.enableMulti;
    /**
     * @type {?}
     * @private
     */
    FFileAdaptDownloadFileDirective.prototype.previewSer;
}
