// @farris/extend-fileupload-adapt-unifile/esm5/lib/ffileupload-adapt-unifile.config.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ffileupload-adapt-unifile.config.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional, InjectionToken, Inject, } from '@angular/core';
import { FARRIS_UPLOADSERVICE_LOCALEDATA } from './locale';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function FFileUploadAdaptUnifileConfig() { }
if (false) {
    /** @type {?|undefined} */
    FFileUploadAdaptUnifileConfig.prototype.rootId;
    /** @type {?|undefined} */
    FFileUploadAdaptUnifileConfig.prototype.formId;
    /** @type {?|undefined} */
    FFileUploadAdaptUnifileConfig.prototype.mode;
    /** @type {?|undefined} */
    FFileUploadAdaptUnifileConfig.prototype.options;
}
/** @type {?} */
export var FFileUploadAdaptUnifileConfigToken = new InjectionToken('MFFileUploadAdaptUnifileConfig');
var FFileUploadAdaptUnifileConfigService = /** @class */ (function () {
    function FFileUploadAdaptUnifileConfigService(exConfig) {
        this.config = (/** @type {?} */ ({
            rootId: '',
            formId: '',
            mode: 0
        }));
        this.basePerfixStr = '';
        if (exConfig) {
            Object.assign(this.config, exConfig);
        }
    }
    /**
     * 获取config属性的值
     */
    /**
     * 获取config属性的值
     * @return {?}
     */
    FFileUploadAdaptUnifileConfigService.prototype.getConfig = /**
     * 获取config属性的值
     * @return {?}
     */
    function () {
        return this.config;
    };
    /**
     * 设置config的属性值
     * @param key
     * @param value
     */
    /**
     * 设置config的属性值
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    FFileUploadAdaptUnifileConfigService.prototype.setConfig = /**
     * 设置config的属性值
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        this.config[key] = value;
    };
    /**
     * 设置基础路径
     * @param value
     */
    /**
     * 设置基础路径
     * @param {?} value
     * @return {?}
     */
    FFileUploadAdaptUnifileConfigService.prototype.setBasePath = /**
     * 设置基础路径
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.basePerfixStr = value;
    };
    /**
     * 获取基础路径
     * @returns
     */
    /**
     * 获取基础路径
     * @return {?}
     */
    FFileUploadAdaptUnifileConfigService.prototype.getBasePath = /**
     * 获取基础路径
     * @return {?}
     */
    function () {
        return this.basePerfixStr;
    };
    /**
     * @param {?} localSer
     * @return {?}
     */
    FFileUploadAdaptUnifileConfigService.prototype.setLocalState = /**
     * @param {?} localSer
     * @return {?}
     */
    function (localSer) {
        if (this.localSerStorage) {
            return true;
        }
        if (localSer && !this.localSerStorage) {
            localSer.setLocaleData(FARRIS_UPLOADSERVICE_LOCALEDATA);
            this.localSerStorage = localSer;
            return true;
        }
        return false;
    };
    /**
     * 获取多语言数据
     * 某些组件没有重新生成，要支持多语言
     * @param key
     */
    /**
     * 获取多语言数据
     * 某些组件没有重新生成，要支持多语言
     * @param {?} lanKey
     * @return {?}
     */
    FFileUploadAdaptUnifileConfigService.prototype.getLocalStr = /**
     * 获取多语言数据
     * 某些组件没有重新生成，要支持多语言
     * @param {?} lanKey
     * @return {?}
     */
    function (lanKey) {
        /** @type {?} */
        var result = "";
        /** @type {?} */
        var defaultLocalId = 'en';
        if (this.localSerStorage) {
            result = this.localSerStorage.getValue(lanKey);
        }
        else {
            defaultLocalId = localStorage.getItem('languageCode') || 'zh-CHS';
        }
        if (!result) {
            result = lanKey.split('.').reduce((/**
             * @param {?} obj
             * @param {?} key
             * @return {?}
             */
            function (obj, key) {
                if (obj) {
                    return obj[key];
                }
                else {
                    return null;
                }
            }), FARRIS_UPLOADSERVICE_LOCALEDATA[defaultLocalId]);
        }
        return result;
    };
    FFileUploadAdaptUnifileConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    FFileUploadAdaptUnifileConfigService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [FFileUploadAdaptUnifileConfigToken,] }] }
    ]; };
    /** @nocollapse */ FFileUploadAdaptUnifileConfigService.ngInjectableDef = i0.defineInjectable({ factory: function FFileUploadAdaptUnifileConfigService_Factory() { return new FFileUploadAdaptUnifileConfigService(i0.inject(FFileUploadAdaptUnifileConfigToken, 8)); }, token: FFileUploadAdaptUnifileConfigService, providedIn: "root" });
    return FFileUploadAdaptUnifileConfigService;
}());
export { FFileUploadAdaptUnifileConfigService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FFileUploadAdaptUnifileConfigService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    FFileUploadAdaptUnifileConfigService.prototype.basePerfixStr;
    /**
     * @type {?}
     * @private
     */
    FFileUploadAdaptUnifileConfigService.prototype.localSerStorage;
}
;
