// @farris/external-integration/esm2015/lib/component/external-integration/external-integration-service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component/external-integration/external-integration-service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from "@angular/common/http";
import { Injectable, Output, EventEmitter } from "@angular/core";
import { VariableParseService, ViewModel } from "@farris/devkit";
export class ExternalIntegrationService {
    /**
     * @param {?} http
     * @param {?} variableParseService
     * @param {?} vm
     */
    constructor(http, variableParseService, vm) {
        this.http = http;
        this.variableParseService = variableParseService;
        this.vm = vm;
        this.sapData = new EventEmitter();
        this.sapUrl = "/api/runtime/icc/v1.0/externalServiceOpenApi/invokeService";
    }
    /**
     * 调用外部服务
     * @param {?} externalParams
     * @param {?} serviceCode
     * @return {?}
     */
    queryExtIntegrationService(externalParams, serviceCode) {
        /** @type {?} */
        let req = this.assembleSapRequest(externalParams, serviceCode);
        this.http.post(this.sapUrl, req).subscribe((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            /** @type {?} */
            let respObj = Object.create(response);
            if (!respObj || !respObj.success) {
                return;
            }
            /** @type {?} */
            let resp = Object.create(respObj);
            /** @type {?} */
            let sapData = resp.data;
            this.sapData.emit(sapData);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.log(error);
            this.sapData.emit({});
        }));
    }
    /**
     * 组装sap api入参
     * @private
     * @param {?} externalParams
     * @param {?} serviceCode
     * @return {?}
     */
    assembleSapRequest(externalParams, serviceCode) {
        /** @type {?} */
        const params = JSON.parse(JSON.stringify(externalParams));
        /** @type {?} */
        let requestParam = {
            code: serviceCode,
            params: this.variableParseService.parse(params, this.vm.frameContext),
        };
        return requestParam;
    }
}
ExternalIntegrationService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ExternalIntegrationService.ctorParameters = () => [
    { type: HttpClient },
    { type: VariableParseService },
    { type: ViewModel }
];
ExternalIntegrationService.propDecorators = {
    sapData: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    ExternalIntegrationService.prototype.sapData;
    /** @type {?} */
    ExternalIntegrationService.prototype.sapUrl;
    /**
     * @type {?}
     * @private
     */
    ExternalIntegrationService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    ExternalIntegrationService.prototype.variableParseService;
    /**
     * @type {?}
     * @private
     */
    ExternalIntegrationService.prototype.vm;
}
