// @farris/external-integration/esm2015/lib/component/external-integration/grid-external-integration.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component/external-integration/grid-external-integration.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Injector, Renderer2, ViewChild, } from "@angular/core";
import { RuntimeStateService } from "@farris/ui-common";
import { GRID_EDITORS } from "@farris/ui-datagrid";
import { DatagridBaseEditorDirective } from "@farris/ui-datagrid-editors";
import { ExternalIntegrationComponent } from "./external-integration.component";
/** @type {?} */
const ExtIntgrtnDefautOption = {
    readonly: false,
    // 是否只读
    placeholder: "请选择",
    viewType: "tag"
};
export class DatagridExtIntgrtnComponent extends DatagridBaseEditorDirective {
    /**
     * @param {?} render
     * @param {?} el
     * @param {?} rts
     * @param {?} injector
     */
    constructor(render, el, rts, injector) {
        super(render, el, injector);
        this.rts = rts;
        this.stopPropagation = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.inputElement = this.instance.input.nativeElement;
        this.options = Object.assign({}, ExtIntgrtnDefautOption, this.options);
        this.column.editor.options.onBlur = (/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            e.editorRef.startPending();
            e.editorRef.formControl.setValue(this.instance.fieldText);
            this.instance.callSapService();
            this.instance.afterOnblur.subscribe((/**
             * @return {?}
             */
            () => {
                e.editorRef.endPending();
            }));
        });
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        super.ngAfterViewInit();
    }
}
DatagridExtIntgrtnComponent.decorators = [
    { type: Component, args: [{
                selector: "grid-external-integration",
                template: `
    <div [formGroup]="group" class="f-datagrid-cell-formgroup farris-group-auto">
      <web-external-integration
        #extIntgrtn
        style="width: 100%"
        extIntegration-dataMapping
        [placeholder]="options.placeholder"
        [readonly]="options.readonly"
        [mapFields]="options.mapFields"
        [serviceCode] = "options.serviceCode"
        [externalParams] = "options.externalParams"
        [extTableSchemas] = "options.extTableSchemas"
        [formControlName]="column.field"
        [mappingType] = "'grid'"
        
      ></web-external-integration>
    </div>
  `
            }] }
];
/** @nocollapse */
DatagridExtIntgrtnComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: RuntimeStateService },
    { type: Injector }
];
DatagridExtIntgrtnComponent.propDecorators = {
    instance: [{ type: ViewChild, args: ["extIntgrtn",] }]
};
if (false) {
    /** @type {?} */
    DatagridExtIntgrtnComponent.prototype.instance;
    /** @type {?} */
    DatagridExtIntgrtnComponent.prototype.stopPropagation;
    /**
     * @type {?}
     * @private
     */
    DatagridExtIntgrtnComponent.prototype.rts;
}
/** @type {?} */
export const ExtIntgrtnDataGridEditorProvider = {
    provide: GRID_EDITORS,
    useValue: { name: "external-integration", value: DatagridExtIntgrtnComponent },
    multi: true
};
