// @edp-aif/runtime-api/esm5/lib/aif-manual-creation/form-args/FormCreationRuleExecutorArgs.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
/**
 * 表单手工生单时执行生单规则参数类。
 */
var /**
 * 表单手工生单时执行生单规则参数类。
 */
FormCreationRuleExecutorArgs = /** @class */ (function (_super) {
    tslib_1.__extends(FormCreationRuleExecutorArgs, _super);
    function FormCreationRuleExecutorArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * 来源单据ID树形列表
         */
        _this.sourceVoEntityDatas = new Array();
        /**
         * 目标单据主表ID列表
         */
        _this.targetVoDataIds = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormCreationRuleExecutorArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.sourceVoEntityDatas = new Array();
        this.creationRuleId = null;
        this.bizFlowchartId = null;
        this.targetVoDataIds = new Array();
    };
    /**
     * @return {?}
     */
    FormCreationRuleExecutorArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntityDatas", BpmnModelJsonHelper.StringListToJsonArray(this.sourceVoEntityDatas));
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoDataIds", BpmnModelJsonHelper.StringListToJsonArray(this.targetVoDataIds));
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormCreationRuleExecutorArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.sourceVoEntityDatas = BpmnModelJsonHelper.StringListFromJsonArray(BpmnModelJsonHelper.GetAttributeValue(obj, "sourceVoEntityDatas"));
        this.targetVoDataIds = BpmnModelJsonHelper.StringListFromJsonArray((/** @type {?} */ (BpmnModelJsonHelper.GetAttributeValue(obj, "targetVoDataIds"))));
    };
    return FormCreationRuleExecutorArgs;
}(JsonConvertableImpl));
/**
 * 表单手工生单时执行生单规则参数类。
 */
export { FormCreationRuleExecutorArgs };
if (false) {
    /**
     * 生单规则ID
     * @type {?}
     */
    FormCreationRuleExecutorArgs.prototype.creationRuleId;
    /**
     * 业务场景图ID
     * @type {?}
     */
    FormCreationRuleExecutorArgs.prototype.bizFlowchartId;
    /**
     * 来源单据ID树形列表
     * @type {?}
     */
    FormCreationRuleExecutorArgs.prototype.sourceVoEntityDatas;
    /**
     * 目标单据主表ID列表
     * @type {?}
     */
    FormCreationRuleExecutorArgs.prototype.targetVoDataIds;
}
