// @edp-aif/runtime-api/esm5/lib/aif-manual-creation/form-args/FormGetCreationRuleEntityExArgs.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from '@edp-aif/common-api';
var FormGetCreationRuleEntityExArgs = /** @class */ (function (_super) {
    tslib_1.__extends(FormGetCreationRuleEntityExArgs, _super);
    function FormGetCreationRuleEntityExArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormGetCreationRuleEntityExArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.bizFlowchartId = null;
        this.sourceVoId = null;
        this.targetVoId = null;
    };
    /**
     * @return {?}
     */
    FormGetCreationRuleEntityExArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoId", this.targetVoId);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormGetCreationRuleEntityExArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.sourceVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
        this.targetVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoId");
    };
    return FormGetCreationRuleEntityExArgs;
}(JsonConvertableImpl));
export { FormGetCreationRuleEntityExArgs };
if (false) {
    /** @type {?} */
    FormGetCreationRuleEntityExArgs.prototype.bizFlowchartId;
    /** @type {?} */
    FormGetCreationRuleEntityExArgs.prototype.sourceVoId;
    /** @type {?} */
    FormGetCreationRuleEntityExArgs.prototype.targetVoId;
}
