// @gsp-wf/rtdevkit/esm2015/lib/runtime/services/task.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
export class TaskService {
    /**
     * @param {?} httpSvc
     */
    constructor(httpSvc) {
        this.httpSvc = httpSvc;
    }
    /**
     * 获取工作项操作列表
     * @param {?} workItemId
     * @return {?}
     */
    getWorkItemActions(workItemId) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/${workItemId}/actions`;
        return this.httpSvc.get(url);
    }
    /**
     * 执行审批操作
     * @param {?} taskId 任务Id
     * @param {?} context 执行审批上下文
     * @return {?}
     */
    excuteTask(taskId, context) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/complete`;
        context.workItemId = taskId;
        /** @type {?} */
        const options = {
            excuteParam: context
        };
        return this.httpSvc.post(url, context);
    }
    /**
     * 获取跳转活动信息
     * @param {?} workItemId 工作项Id
     * @param {?} action 动作编号
     * @return {?}
     */
    getDestinationScope(workItemId, action) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/${workItemId}/nextNodes/${action}`;
        return this.httpSvc.get(url);
    }
    /**
     * 获取动作实体进而获取web实现和前后事件
     * @param {?} taskId 任务ID
     * @param {?} action 动作编号
     * @return {?}
     */
    getActionEntity(taskId, action) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/approvalAction?code=${action}`;
        return this.httpSvc.get(url);
    }
    /**
     * 执行加签操作
     * @param {?} context 加签上下文
     * @return {?}
     */
    addSign(context) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/counterSignAdd`;
        /** @type {?} */
        const body = {
            excuteParam: context
        };
        return this.httpSvc.post(url, context);
    }
    /**
     * 查询单据信息
     * @param {?} workItemId 工作项Id
     * @return {?}
     */
    getAppInfo(workItemId) {
        /** @type {?} */
        const formParUrl = `/api/runtime/wf/v1.0/workItems/${workItemId}/formInfo`;
        return this.httpSvc.get(formParUrl);
    }
    //#region 查看流程
    /**
     * 基于工作项ID获取bpmnModel
     * @param {?} workitemId 工作项id
     * @return {?}
     */
    getBpmnModelbyWorkItemId(workitemId) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/${workitemId}/bpmnModel`;
        return this.httpSvc.get(url);
    }
    /**
     * 基于工作项Id获取所有活动实例
     * @param {?} id taskId
     * @return {?}
     */
    getAllActiInstsbyTaskId(id) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/activityInstanceSlims?workItemId=${id}`;
        return this.httpSvc.get(url);
    }
    /**
     * 基于活动实例Id获取日志
     * @param {?} actInstId 活动实例Id
     * @return {?}
     */
    getActivityInstanceLogs(actInstId) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/activityInstanceLogs?activityInstanceId=${actInstId}`;
        return this.httpSvc.get(url);
    }
    /**
     * 获取审批日志信息
     * @param {?} taskId 工作项Id
     * @return {?}
     */
    getBizlogByTaskId(taskId) {
        /** @type {?} */
        const logUrl = `/api/runtime/wf/v1.0/workItems/${taskId}/fullComments`;
        // if (false === environment.production) {
        //     logUrl = '/assets/mock/taskbizlog.json';
        // }
        return this.httpSvc.get(logUrl);
    }
    //#endregion
    /**
     * 获取指派信息
     * @param {?} taskId 工作项Id
     * @return {?}
     */
    getAssigneeInfo(taskId) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/${taskId}/assigneeInfo`;
        return this.httpSvc.get(url);
    }
    /**
     * 根据流程实例Id批量指派（提交审批或执行审批后指派弹框调用，_后续调整与工作项的一致_）
     * @param {?} assigneesPa 指派参数
     * @return {?}
     */
    assignNextUsersByProcInstId(assigneesPa) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/processInstances/assignNextUsers`;
        /** @type {?} */
        const body = {
            assigneeInfoParameter: assigneesPa
        };
        return this.httpSvc.post(url, assigneesPa);
    }
    /**
     * 根据工作项ID指派任务
     * @param {?} assigneesPa 指派参数
     * @return {?}
     */
    assignNextUsersByWorkItemId(assigneesPa) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/assignUser`;
        /** @type {?} */
        const body = {
            assigneeInfoParameter: assigneesPa
        };
        return this.httpSvc.post(url, assigneesPa);
    }
    /**
     * 取回
     * @param {?} context
     * @return {?}
     */
    retrieve(context) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/retrieve`;
        /** @type {?} */
        const body = {
            excuteParam: context
        };
        return this.httpSvc.post(url, context);
    }
    /**
     * 任务移交
     * @param {?} transferContext
     * @return {?}
     */
    transfer(transferContext) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/transfer`;
        /** @type {?} */
        const body = {
            excuteParam: transferContext
        };
        return this.httpSvc.post(url, transferContext);
    }
    //#region  任务领用
    /**
     * 领用任务
     * @param {?} taskId 工作项Id
     * @return {?}
     */
    claim(taskId) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/claim`;
        /** @type {?} */
        const options = {
            workItemId: taskId
        };
        return this.httpSvc.post(url, options);
    }
    /**
     * 取消领用
     * @param {?} taskId 任务Id
     * @return {?}
     */
    unClaim(taskId) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/workItems/unclaim`;
        /** @type {?} */
        const options = {
            workItemId: taskId
        };
        return this.httpSvc.post(url, options);
    }
}
TaskService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
TaskService.ctorParameters = () => [
    { type: HttpService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaskService.prototype.httpSvc;
}
