// @gsp-wf/rtdevkit/esm2015/lib/runtime/services/user.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { SessionService, HttpService } from '@ecp-caf/caf-common';
import { forkJoin } from 'rxjs';
export class UserService {
    /**
     * @param {?} http
     * @param {?} httpSvc
     * @param {?} sessionSvc
     */
    constructor(http, httpSvc, sessionSvc) {
        this.http = http;
        this.httpSvc = httpSvc;
        this.sessionSvc = sessionSvc;
    }
    /**
     * @param {?} ids
     * @return {?}
     */
    getUsersByIds(ids) {
        /** @type {?} */
        var pathUrl = '/api/runtime/sys/v1.0/users';
        return forkJoin(ids.map((/**
         * @param {?} id
         * @return {?}
         */
        id => this.http.get(pathUrl + `/${id}` + `?mode=0002`)))); //mode用于获取组织全路径
    }
    /**
     * @param {?} searchValue
     * @param {?} index
     * @param {?} size
     * @return {?}
     */
    getUserList(searchValue, index, size) {
        if (searchValue) {
            return this.getUser('codeorname', searchValue, index, size);
        }
        else {
            return this.getUser('', '', index, size);
        }
    }
    /**
     * @private
     * @param {?} field
     * @param {?} value
     * @param {?} index
     * @param {?} size
     * @return {?}
     */
    getUser(field, value, index, size) {
        /** @type {?} */
        let url = `/api/runtime/sys/v1.0/users`;
        url += field ?
            `?param={"notGetCurrentUser":false,"pageIndex":${index},"pageSize":${size},"${field}":'${value}',"orderField":"order", "withAllOrg": "true", "queryMode":"0002"}` :
            `?param={"notGetCurrentUser":false,"pageIndex":${index},"pageSize":${size},"orderField":"order", "withAllOrg": "true", "queryMode":"0002"}`;
        return this.httpSvc.get(url);
    }
    /**
     * @param {?} searchValue
     * @param {?} index
     * @param {?} size
     * @param {?=} withAllOrg
     * @param {?=} orgId
     * @return {?}
     */
    getUserList_Assign(searchValue, index, size, withAllOrg, orgId) {
        if (searchValue) {
            return this.getUser_Assign('codeorname', searchValue, index, size, withAllOrg, orgId);
        }
        else {
            return this.getUser_Assign('', '', index, size, withAllOrg, orgId);
        }
    }
    /**
     * @private
     * @param {?} field
     * @param {?} value
     * @param {?} index
     * @param {?} size
     * @param {?} withAllOrg
     * @param {?} orgId
     * @return {?}
     */
    getUser_Assign(field, value, index, size, withAllOrg, orgId) {
        /** @type {?} */
        let url = `/api/runtime/sys/v1.0/users`;
        url += field ?
            `?param={"notGetCurrentUser":false,"pageIndex":${index},"pageSize":${size},"${field}":'${value}',"orderField":"order", "withAllOrg": "true", "queryMode":"0002"}` :
            `?param={"notGetCurrentUser":false,"pageIndex":${index},"pageSize":${size},"orderField":"order", "withAllOrg":'${withAllOrg}',"orgId":'${orgId}', "queryMode":"0002"}`;
        return this.httpSvc.get(url);
    }
    /**
     * 构造Header
     * @private
     * @return {?}
     */
    getBatchHeader() {
        /** @type {?} */
        const sessionId = this.sessionSvc.getCurrentSeesionId();
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', sessionId).set('Accept', 'application/json');
        // 此处key必须是headers
        return headers;
    }
    /**
     * 转换配置大小写
     * @param {?} jsonObj Object
     * @return {?}
     */
    toUpperCase(jsonObj) {
        if (typeof (jsonObj) === 'object') {
            for (const key of Object.keys(jsonObj)) {
                jsonObj[key.substring(0, 1).toUpperCase() + key.substring(1)] = jsonObj[key];
                delete (jsonObj[key]);
            }
        }
        return jsonObj;
    }
}
UserService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
UserService.ctorParameters = () => [
    { type: HttpClient },
    { type: HttpService },
    { type: SessionService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    UserService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    UserService.prototype.httpSvc;
    /**
     * @type {?}
     * @private
     */
    UserService.prototype.sessionSvc;
}
