// @gsp-sys/rtf-apphelp/esm5/lib/rtf-app-help/rtf-app-help.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { GridModule } from '@progress/kendo-angular-grid';
import { LayoutModule } from '@progress/kendo-angular-layout';
import { TreeViewModule } from '@progress/kendo-angular-treeview';
import { AppModuleTreeComponent } from './app-module-tree/app-module-tree.component';
import { RtfAppComponent } from './rtf-app/rtf-app.component';
import { RtfAppHelpComponent } from './rtf-app-help/rtf-app-help.component';
import { RtfAppService } from './service/rtf-app.service';
import { RtfUIWebapiService } from '../common/rtf-ui-webapi.service';
import { HttpClientModule } from '@angular/common/http';
import { ModalModule } from '@farris/ui-modal';
import { MessagerModule } from '@farris/ui-messager';
import { I18NresourceService } from './service/i18N/i18-nresource.service';
var RtfAppHelpModule = /** @class */ (function () {
    function RtfAppHelpModule() {
    }
    RtfAppHelpModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        HttpClientModule,
                        // kendo
                        GridModule,
                        LayoutModule,
                        TreeViewModule,
                        // farris
                        ModalModule.forRoot(),
                        MessagerModule.forRoot(),
                    ],
                    declarations: [
                        AppModuleTreeComponent,
                        RtfAppComponent,
                        RtfAppHelpComponent,
                    ],
                    entryComponents: [
                        RtfAppComponent
                    ],
                    providers: [
                        RtfUIWebapiService,
                        RtfAppService,
                        I18NresourceService
                    ],
                    exports: [
                        RtfAppComponent,
                        RtfAppHelpComponent
                    ]
                },] }
    ];
    return RtfAppHelpModule;
}());
export { RtfAppHelpModule };
