// @farris/mobile-command-services/esm5/lib/entity-services/entity.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity-services/entity.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EntityTraversingService } from './entity-traversing.service';
import { EntityManipulationService } from './entity-manipulation.service';
import { EntityAggregationService } from './entity-aggregation.service';
/**
 * 实体服务
 * \@scope FrameComponent
 */
var /**
 * 实体服务
 * \@scope FrameComponent
 */
EntityService = /** @class */ (function () {
    function EntityService(viewModelContext) {
        this.viewModelContext = viewModelContext;
        this.traversingService = new EntityTraversingService(viewModelContext);
        this.manipulationService = new EntityManipulationService(viewModelContext);
        this.aggregationService = new EntityAggregationService(viewModelContext);
    }
    /**
     * 获取path对应的属性值
     */
    /**
     * 获取path对应的属性值
     * @param {?} path
     * @return {?}
     */
    EntityService.prototype.getPropValue = /**
     * 获取path对应的属性值
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return this.traversingService.getPropValue(path);
    };
    /**
     * 设置path对应的属性值
     */
    /**
     * 设置path对应的属性值
     * @param {?} path
     * @param {?} value
     * @return {?}
     */
    EntityService.prototype.setPropValue = /**
     * 设置path对应的属性值
     * @param {?} path
     * @param {?} value
     * @return {?}
     */
    function (path, value) {
        return this.manipulationService.setPropValue(path, value);
    };
    /**
     * 设置path对应的Entity数据
     */
    /**
     * 设置path对应的Entity数据
     * @param {?} path
     * @return {?}
     */
    EntityService.prototype.getEntityData = /**
     * 设置path对应的Entity数据
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return this.traversingService.getEntityData(path);
    };
    /**
     * 设置path对应的EntityList数据
     */
    /**
     * 设置path对应的EntityList数据
     * @param {?} path
     * @return {?}
     */
    EntityService.prototype.getEntityListData = /**
     * 设置path对应的EntityList数据
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return this.traversingService.getEntityListData(path);
    };
    /**
     * 获取path对应EntityList记录数
     */
    /**
     * 获取path对应EntityList记录数
     * @param {?} path
     * @return {?}
     */
    EntityService.prototype.count = /**
     * 获取path对应EntityList记录数
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return this.aggregationService.count(path);
    };
    /**
     * 获取path对应属性的合计值
     */
    /**
     * 获取path对应属性的合计值
     * @param {?} path
     * @return {?}
     */
    EntityService.prototype.sum = /**
     * 获取path对应属性的合计值
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return this.aggregationService.sum(path);
    };
    /**
     * 获取path对应属性的平均值
     */
    /**
     * 获取path对应属性的平均值
     * @param {?} path
     * @return {?}
     */
    EntityService.prototype.avg = /**
     * 获取path对应属性的平均值
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return this.aggregationService.avg(path);
    };
    /**
     * 获取path对应属性的最大值
     */
    /**
     * 获取path对应属性的最大值
     * @param {?} path
     * @return {?}
     */
    EntityService.prototype.max = /**
     * 获取path对应属性的最大值
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return this.aggregationService.max(path);
    };
    /**
     * 获取path对应属性的最小值
     */
    /**
     * 获取path对应属性的最小值
     * @param {?} path
     * @return {?}
     */
    EntityService.prototype.min = /**
     * 获取path对应属性的最小值
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return this.aggregationService.min(path);
    };
    return EntityService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    EntityService.prototype.traversingService;
    /**
     * @type {?}
     * @private
     */
    EntityService.prototype.manipulationService;
    /**
     * @type {?}
     * @private
     */
    EntityService.prototype.aggregationService;
    /**
     * @type {?}
     * @private
     */
    EntityService.prototype.viewModelContext;
}
export { EntityService };
