// @farris/ui-organization-selector/esm2015/lib/organization-selector.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/organization-selector.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, forwardRef, HostBinding, Injector, Input, NgZone, Output, Renderer2, ViewChild, ChangeDetectorRef } from '@angular/core';
import { NgControl, NG_VALUE_ACCESSOR } from '@angular/forms';
import { DialogComponent } from '@farris/ui-dialog';
import { ApiService } from '../utils/api.service';
import { CommonUtil } from '../utils/common.tool';
import { FavoritesTool } from '../utils/favorites.tool';
import { ChangeType } from '../utils/index.type';
import { StateMgrTool } from '../utils/state.mgr.tool';
import { Subject } from 'rxjs';
import * as _ from 'lodash-es';
import { debounceTime, distinctUntilChanged, switchMap, tap } from 'rxjs/operators';
import { GroupComponent } from './tab-group/group.component';
import { OrganizationComponent } from './tab-organization/to.component';
import { FavoritesComponent } from './tab-favorites/favorites.component';
import { MessagerService } from '@farris/ui-messager';
import { NotifyService } from '@farris/ui-notify';
import { LocaleService } from '@farris/ui-locale';
import { ORGANIZATION_SELECTOR_LOCALE } from '../utils/locale';
export class OrganizationSelectorComponent {
    /**
     * @param {?} render
     * @param {?} injector
     * @param {?} favoritesTool
     * @param {?} apiService
     * @param {?} ngZone
     * @param {?} stateMgr
     */
    constructor(render, injector, favoritesTool, apiService, ngZone, stateMgr) {
        this.render = render;
        this.injector = injector;
        this.favoritesTool = favoritesTool;
        this.apiService = apiService;
        this.ngZone = ngZone;
        this.stateMgr = stateMgr;
        this.baseCls = true;
        this.width = '100%';
        // input框是否显示
        this.showInput = true;
        // 业务标识
        this.busNum = 'common';
        this.multiSelect = true;
        this.placeholder = '请选择';
        this.searchPlaceholder = '请输入编号或名称搜索';
        this.readonly = false;
        // 传入的url
        this.absUserOrgUrl = '/api/runtime/sys/v1.0/absUserOrg';
        this.absRfQueryUrl = '/api/runtime/sys/v1.0/absRecentContacts/query';
        this.absRfModifyUrl = '/api/runtime/sys/v1.0/absRecentContacts/modify';
        this.groupApiUrl = "/api/runtime/sys/v1.0/org-group";
        this.groupExtendUrl = '/platform/runtime/wf/webapp/ng-farris-mobile-ui/extend-org.js';
        /**
         * 数据源类型
         * 默认为： System_organization  系统数据源   获取系统人员、系统组织
         * Administrative_organization  已扩展数据源  获取行政组织、行政人员
         * absOrgType {string}
         * PersonnelSelectorComponent
         */
        this.absOrgType = 'System_organization';
        // 启用分层加载
        this.enableHierarchicalLoading = true;
        // 不包含停用组织
        this.notGetStopOrg = true;
        this.primaryField = 'orgId';
        this.displayField = 'name';
        this.orgIds = [];
        // 展开层级
        this.layer = 1;
        // 是否启用级联选中
        this.cascadeCheck = true;
        // 级联选中下级
        this.cascadeDown = true;
        // 级联选中上级
        this.cascadeUp = false;
        // 显示全选按钮
        this.showCheckAll = false;
        this.keepSelect = false;
        // 一次性最多选中多少条
        this.selectedLimit = 50;
        this.showTabIds = 'organization,favorites';
        this.activeTabId = 'organization';
        // 搜索前端分页50条
        this.searchPageSize = 50;
        /**
         * 多选分隔符
         */
        this.multipleChoiceSeparator = ',';
        this.draggable = true;
        this.virtualizedLoad = false;
        this.dataPermission = false;
        this.dataPermissionIds = [];
        this.filterType = 'none';
        this.selectionsChange = new EventEmitter();
        // 全部清除按钮触发回调
        this.inputClear = new EventEmitter();
        // 标签删除按钮出发回调
        this.tagRemoved = new EventEmitter();
        this.viewType = 'tag'; // 默认为tag标签形式，可支持text，后续改为input类型属性 
        // 默认为tag标签形式，可支持text，后续改为input类型属性 
        this.selections = [];
        this.originalSelections = [];
        this.finalText = [];
        this.lookupIcon = '<i class="f-icon f-icon-lookup"></i>';
        this.searchIcon = '<span class="k-icon k-i-search"></span>';
        this.organizationTabId = 'organization';
        this.groupTabId = 'group';
        this.favoritesTabId = 'favorites';
        this.dialogHeight = 548;
        this.dialogId = 'dialogId';
        this.isShow = false;
        this.confirmToken = false;
        this.hideSearchPanel = true;
        this.searchText = '';
        this.groupDialogOpend = false;
        this.groupDialogTitle = '';
        this.validGroupTitle = true;
        this.groupTitleStream = new Subject();
        this.selectionsShow = false;
        this.orgTabActive = false;
        this.groupTabActive = false;
        this.againstSingleSelect = false;
        this.localeService = null;
        this.languagesArray = [
            { code: 'zh-CHS', name: '中文简体', default: true },
            { code: 'en', name: 'English' },
            { code: 'zh-CHT', name: '中文繁体' }
        ];
        this._disableOrgList = [];
        this.farrisDialogbeforeClose = (/**
         * @return {?}
         */
        () => {
            this.isShow = false;
            if (this.confirmToken) {
                this.originalSelections = _.cloneDeep(this.selections);
            }
            else {
                this.selections = _.cloneDeep(this.originalSelections);
            }
            if (!this.hideSearchPanel || this.searchText) {
                this.searchText = null;
                this.clearSearch();
            }
            this.confirmToken = false;
            return true;
        });
        this.onModelChange = (/**
         * @param {?=} val
         * @return {?}
         */
        (val) => { });
        this.onModelTouched = (/**
         * @param {?=} val
         * @return {?}
         */
        (val) => { });
        this.apiService.absUrl = this.absUserOrgUrl;
        this.apiService.groupApiUrl = this.groupApiUrl;
        this.cdRef = this.injector.get(ChangeDetectorRef);
        this.messageService = this.injector.get(MessagerService, null);
        this.notifyService = this.injector.get(NotifyService, null);
        this.localeService = this.injector.get(LocaleService);
        if (!this.localeService.langData['organizationSelector']) {
            this.localeService.setLocaleData(ORGANIZATION_SELECTOR_LOCALE);
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set disableOrgList(value) {
        if (this.filterType == 'disable') {
            if (typeof value == 'object' && Array.isArray(value)) {
                this._disableOrgList = value;
            }
            else if (typeof value == 'string') {
                this._disableOrgList = value.split(',');
            }
        }
    }
    /**
     * @return {?}
     */
    get disableOrgList() {
        return this._disableOrgList;
    }
    // 自定义过滤feature：不可选中组织，历史属性兼容保留
    /**
     * @param {?} value
     * @return {?}
     */
    set filterOrgList(value) {
        if (typeof value == 'object' && Array.isArray(value)) {
            this._disableOrgList = value;
        }
        else if (typeof value == 'string') {
            this._disableOrgList = value.split(',');
        }
        this.filterType = this._disableOrgList.length ? 'disable' : this.filterType;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    writeValue(obj) {
        if (this.viewType == 'tag') {
            if (!!obj && typeof obj === 'string') {
                this.finalText = obj.split(this.multipleChoiceSeparator);
            }
            else {
                this.finalText = [];
            }
        }
        else if (this.viewType == 'text') {
            this.finalText = obj;
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onModelChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onModelTouched = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        throw new Error('Method not implemented.');
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.initDialogHeight();
        this.listenSelectionsChange();
        this.initTabs();
        this.initGroupSearch();
        this.ngControl = this.injector.get(NgControl, null);
        this.mapFields && this.mapFields2Component();
        this.viewType = this.dataPermission ? 'text' : 'tag';
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.listenWindowResize();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const orgIdsChange = changes.orgIds;
        if (orgIdsChange && orgIdsChange.currentValue) {
            this.assembleDataByOrgIds(orgIdsChange.currentValue);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.selectionsSub.unsubscribe();
    }
    /**
     * @private
     * @param {?} selections
     * @param {?=} changeType
     * @param {?=} from
     * @return {?}
     */
    emitSelections(selections, changeType = ChangeType.ValueChanged, from = 'form') {
        this.stateMgr.emitSelections(selections, changeType, from);
    }
    // 根据字段映射关系获取当前已选组织的id，进而在打开选组织弹窗时进行相应展示
    /**
     * @private
     * @return {?}
     */
    mapFields2Component() {
        /** @type {?} */
        const helpFields = Object.keys(this.mapFields);
        if (!helpFields.includes(this.primaryField)) {
            // this.notifyService.warning(this.localeService.getValue("organizationSelector.errorMessage.mapFields"));
            return;
        }
        if (this.ngControl &&
            this.ngControl.formDirective &&
            this.ngControl.formDirective.form &&
            this.ngControl.formDirective.form.bindingData) {
            /** @type {?} */
            const bindingData = this.ngControl.formDirective.form.bindingData;
            if (bindingData.setValue) {
                /** @type {?} */
                const bindingPath = this.ngControl.formDirective.form.bindingPath;
                /** @type {?} */
                let pathArr = [];
                if (bindingPath) {
                    pathArr = bindingPath.split('/').filter((/**
                     * @param {?} n
                     * @return {?}
                     */
                    n => n !== ''));
                }
                /** @type {?} */
                const multiPath = this.mapFields[this.primaryField].split(this.multipleChoiceSeparator);
                /** @type {?} */
                const fieldPaths = pathArr.concat(multiPath[0].split('.'));
                /** @type {?} */
                const ids = bindingData.getValue(fieldPaths);
                if (!ids) {
                    this.emitSelections([]);
                    return;
                }
                this.assembleDataByOrgIds(ids.split(this.multipleChoiceSeparator));
            }
        }
    }
    // 根据传入的组织id数组获取对应的组织
    // 由ngOnChanges()、mapFields2Component()调用
    /**
     * @private
     * @param {?} ids
     * @return {?}
     */
    assembleDataByOrgIds(ids) {
        if (!ids || ids.length === 0) {
            return;
        }
        if (ids && ids.length > 1 && !this.multiSelect) {
            this.againstSingleSelect = true;
            return;
        }
        /** @type {?} */
        const param = {
            absOrgType: this.absOrgType,
            idList: this.filterType == 'disable' ? ids.filter((/**
             * @param {?} id
             * @return {?}
             */
            id => !this._disableOrgList.includes(id))) : ids,
            getOrgAllPath: true,
            getCurUser: false
        };
        this.apiService.getOrganization(param).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        (res) => {
            if (res && res.data) {
                if (this.absOrgType == 'System_organization') {
                    this.selections = res.data;
                }
                else {
                    this.selections = res.data.filter((/**
                     * @param {?} data
                     * @return {?}
                     */
                    data => ids.find((/**
                     * @param {?} id
                     * @return {?}
                     */
                    id => id == data[this.primaryField]))));
                }
                if (!this.mapFields) {
                    this.assembleFinalText();
                    if (this.dataPermission) {
                        this.emitSelections(this.selections);
                    }
                }
                else {
                    this.emitSelections(this.selections);
                }
            }
        }));
    }
    // 构造回显至input文本框的内容
    // 在有mapFields的前提下，由assembleDataByOrgIds()调用，clearAll()、removeItem()、returnResult()也会调用
    /**
     * @private
     * @return {?}
     */
    assembleFinalText() {
        this.finalText = [];
        this.selections.forEach((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        (item, index) => {
            /** @type {?} */
            const formatedName = this.formatFn && this.formatFn(item);
            /** @type {?} */
            const finalText = formatedName ? formatedName : item[this.displayField];
            this.finalText.push(finalText);
        }));
        this.finalText = this.finalText.slice();
    }
    // 初始化弹窗id及高度
    // 由ngOnInit()调用
    /**
     * @return {?}
     */
    initDialogHeight() {
        /** @type {?} */
        const uuid = CommonUtil.newGuid();
        this.dialogId += uuid;
        /** @type {?} */
        const windowInnerHeight = window.innerHeight;
        this.dialogHeight = windowInnerHeight - 60;
        if (this.dialogHeight > 680) {
            this.dialogHeight = 680;
        }
    }
    // 监听selections的变更，变更包含三种形式：valueChange、append、remove
    // 由ngOnInit()调用
    /**
     * @return {?}
     */
    listenSelectionsChange() {
        this.selectionsSub = this.stateMgr.selectionsSubject.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            if (data.changeType === ChangeType.ValueChanged) {
                this.selections = _.cloneDeep(data.value);
            }
            else if (data.changeType === ChangeType.Remove) {
                this.selections = this.selections.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => item[this.primaryField] !== data.value[this.primaryField]));
            }
            else if (data.changeType === ChangeType.Append) {
                CommonUtil.findIndex(data.value[this.primaryField], this.primaryField, this.selections) === -1 ? this.selections.push(data.value) : null;
            }
            this.cdRef.detectChanges();
        }));
    }
    // 根据showTabIds判断是否启用收藏功能、群组功能
    // 由ngOnInit()调用
    /**
     * @private
     * @return {?}
     */
    initTabs() {
        if (!this.showTabIds.includes(this.activeTabId)) {
            this.activeTabId = this.showTabIds.split(',')[0];
        }
        if (this.dataPermission) {
            this.showTabIds = this.organizationTabId;
            this.activeTabId = this.organizationTabId;
            this.orgTabActive = true;
            return;
        }
        this.currentTabId = this.activeTabId;
        if (this.showTabIds.indexOf(this.favoritesTabId) !== -1) {
            this.enableFavorites = true;
        }
        else {
            this.enableFavorites = false;
        }
        if (this.showTabIds.indexOf(this.groupTabId) !== -1) {
            this.enableGroup = true;
        }
        else {
            this.enableGroup = false;
        }
        if (this.activeTabId == this.organizationTabId) {
            this.orgTabActive = true;
        }
        else if (this.activeTabId == this.groupTabId) {
            this.groupTabActive = true;
        }
        // 启用群组功能扩展函数
        /*
        if (this.enableGroup) {
          const ccwork = window.top.document.createElement("script")
          ccwork.src = this.groupExtendUrl + '?vs=' + Math.random();
          window.document.head.appendChild(ccwork);
        }
        */
    }
    // 判断群组名称是否合法
    // 由ngOnInit()调用
    /**
     * @private
     * @return {?}
     */
    initGroupSearch() {
        if (this.enableGroup) {
            this.groupTitleStream.pipe(debounceTime(500), distinctUntilChanged()).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            data => {
                if (!!data) {
                    this.validGroupTitle = true;
                }
                else {
                    this.validGroupTitle = false;
                }
            }));
        }
    }
    // 当浏览器窗口大小变更时，变更弹窗的大小
    // 由ngAfterViewInit()调用
    /**
     * @return {?}
     */
    listenWindowResize() {
        /** @type {?} */
        const dialogElement = document.querySelector('#' + this.dialogId);
        /** @type {?} */
        let targetElement = null;
        if (!!dialogElement) {
            targetElement = dialogElement.querySelector('.modal-dialog');
        }
        window.addEventListener('resize', (/**
         * @return {?}
         */
        () => {
            if (!!targetElement) {
                /** @type {?} */
                let height = window.innerHeight - 60;
                if (height > 680) {
                    height = 680;
                }
                this.dialogHeight = height;
                /** @type {?} */
                const innerHeight = height + 'px';
                /** @type {?} */
                const marginTop = '-' + height / 2 + 'px';
                this.render.setStyle(targetElement, 'height', innerHeight);
                this.render.setStyle(targetElement, 'margin-top', marginTop);
            }
        }));
    }
    // 初始化组织收藏数据
    // 由showFarrisDialog()调用，即打开选组织弹窗时发送请求
    /**
     * @private
     * @return {?}
     */
    initFavorites() {
        this.favoritesTool.absOrgType = this.absOrgType;
        this.favoritesTool.busNum = this.busNum;
        this.favoritesTool.idField = this.primaryField;
        this.favoritesTool.absRfQueryUrl = this.absRfQueryUrl;
        this.favoritesTool.absRfModifyUrl = this.absRfModifyUrl;
        this.favoritesTool.getFavoritesList().subscribe((/**
         * @param {?} result
         * @return {?}
         */
        (result) => {
            /** @type {?} */
            const resData = result ? ((/** @type {?} */ (result))).orgData : [];
            this.favoritesTool.favoritesList = resData;
            this.favoritesTool.setFavoritesList(resData);
        }));
    }
    // 打开选组织弹窗
    /**
     * @return {?}
     */
    showFarrisDialog() {
        this.isShow = true;
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        () => {
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.dialog.show();
                if (this.showTabIds.indexOf(this.favoritesTabId) !== -1) {
                    this.initFavorites();
                }
                if (this.againstSingleSelect) {
                    this.notifyService.warning(this.localeService.getValue("organizationSelector.errorMessage.multiSelect"));
                    return;
                }
                if (this.mapFields && !Object.keys(this.mapFields).includes(this.primaryField)) {
                    this.notifyService.warning(this.localeService.getValue("organizationSelector.errorMessage.mapFields"));
                    return;
                }
                this.stateMgr.emitSelections(this.selections, ChangeType.ValueChanged, 'form');
            }));
        }));
    }
    // input文本框内点击清除按钮时触发
    /**
     * @return {?}
     */
    clearAll() {
        this.selections = [];
        this.assembleFinalText();
        this.onModelChange(this.finalText.join());
        this.onModelTouched(this.finalText.join());
        this.emitSelections([]);
        this.inputClear.emit();
    }
    // input文本框内删除某一个已选组织时触发
    /**
     * @param {?} data
     * @param {?} index
     * @return {?}
     */
    removeItem(data, index) {
        if (this.selections.length > 0) {
            this.selections.splice(index, 1);
            this.assembleFinalText();
            this.emitSelections(data, ChangeType.Remove);
        }
        else {
            this.finalText = this.finalText.filter((/**
             * @param {?} text
             * @return {?}
             */
            text => text !== data));
        }
        this.onModelChange(this.finalText.join());
        this.onModelTouched(this.finalText.join());
        this.tagRemoved.emit({
            deleteItem: data,
            deleteIndex: index,
            selections: this.selections
        });
    }
    // 清空最近已选组织
    /**
     * @return {?}
     */
    clearSeletions() {
        this.stateMgr.emitSelections([], ChangeType.ValueChanged, 'form');
    }
    // 点击搜索按钮或者敲击回车时触发
    /**
     * @return {?}
     */
    onSearch() {
        if (!!this.searchText) {
            this.hideSearchPanel = false;
            this.onSearchToken = CommonUtil.newGuid();
        }
    }
    // 清空组织搜索内容时触发
    /**
     * @return {?}
     */
    clearSearch() {
        this.hideSearchPanel = true;
        this.onSearchToken = false;
        if (this.currentTabId = this.organizationTabId) {
            this.libOrgTree.farrisTree.clearAll();
            this.libOrgTree.farrisTree.rowHeight = 28;
            if (this.multiSelect) {
                this.libOrgTree.checkOrgTreeNodes();
            }
            else {
                this.libOrgTree.selectOrgTreeNode();
            }
        }
    }
    // 监听搜索框内输入的文本
    /**
     * @param {?} e
     * @return {?}
     */
    listenSearchText(e) {
        if (e.code === 'Backspace' || e.code === 'Delete') {
            if (!this.searchText) {
                this.clearSearch();
            }
        }
    }
    // 监听搜索页的关闭事件，目前只会接收到true
    /**
     * @param {?} e
     * @return {?}
     */
    listenClosePanelChange(e) {
        this.hideSearchPanel = e;
        // 关闭搜索结果页后，需要更新树节点的是否选中状态
        if (this.currentTabId == this.organizationTabId) {
            this.libOrgTree.farrisTree.clearAll();
            this.libOrgTree.farrisTree.rowHeight = 28;
            if (this.multiSelect) {
                this.libOrgTree.checkOrgTreeNodes();
            }
            else {
                this.libOrgTree.selectOrgTreeNode();
            }
        }
    }
    // 关闭保存群组的弹窗时点击“取消”或者右上角叉号触发
    /**
     * @return {?}
     */
    closeGroupDialog() {
        this.groupDialogOpend = false;
        this.editingGroupId = null;
        this.groupTitle = null;
        this.groupNote = null;
        this.groupDialogTitle = this.localeService.getValue("organizationSelector.group.saveGroup");
    }
    // 在保存群组的弹窗内，修改群组名称时触发，监听keyup事件
    /**
     * @param {?} e
     * @return {?}
     */
    listenGroupTitle(e) {
        this.groupTitleStream.next(this.groupTitle);
    }
    // 选组织的弹窗内有一个“已选”标志，鼠标移上去后会显示已选组织的页面，该页面内有“保存为群组”的按钮，点击后可以打开保存群组的弹窗
    /**
     * @param {?} e
     * @return {?}
     */
    listenOpenGroup(e) {
        this.groupDialogOpend = e;
        this.groupDialogTitle = this.localeService.getValue("organizationSelector.group.saveGroup");
    }
    // 点击群组编辑按钮触发，可修改选中群组的名称及描述
    /**
     * @param {?} item
     * @return {?}
     */
    listenGroupChange(item) {
        this.groupDialogOpend = true;
        this.editingGroupId = item.id;
        this.groupTitle = item.title;
        this.groupNote = item.description;
        this.groupDialogTitle = this.localeService.getValue("organizationSelector.group.editGroup");
    }
    // 关闭保存群组的弹窗时点击“确定”触发，将当前已选组织保存为一个新的群组
    /**
     * @return {?}
     */
    saveGroup() {
        if (!this.groupTitle) {
            this.validGroupTitle = false;
            return;
        }
        /** @type {?} */
        const groupId = CommonUtil.newGuid();
        /** @type {?} */
        let groupCode = 'orgGroup_private_';
        groupCode += Math.random().toString(16).substring(2, 10);
        /** @type {?} */
        const addIds = [];
        this.selections.forEach((/**
         * @param {?} selection
         * @return {?}
         */
        selection => {
            addIds.push(selection[this.primaryField]);
        }));
        if (addIds.length > 1000) {
            this.messageService.show('warning', this.localeService.getValue("organizationSelector.errorMessage.limit"), true);
            this.groupDialogOpend = false;
            this.groupTitle = null;
            this.groupNote = null;
            return;
        }
        this.groupDialogOpend = false;
        /** @type {?} */
        const body = {
            groupId,
            groupCode,
            title: this.groupTitle,
            description: this.groupNote,
            absOrgType: this.absOrgType,
            changeType: Number(1)
        };
        if (this.editingGroupId) {
            body.groupId = this.editingGroupId;
            body.changeType = Number(2);
            delete body.groupCode;
            this.updateNameOrNote(body);
            return;
        }
        /** @type {?} */
        const newGroup$ = this.apiService.saveGroup(body);
        newGroup$.pipe(switchMap((/**
         * @param {?} res
         * @return {?}
         */
        (res) => {
            if (res) {
                this.libOrgGroup.selectedGroupId = groupId;
                this.groupTitle = null;
                this.groupNote = null;
                /** @type {?} */
                const addOrgsInNewGroup$ = this.apiService.saveGroup({ groupId, addIds, deleteIds: [], changeType: Number(3) });
                return addOrgsInNewGroup$.pipe(tap((/**
                 * @param {?} data
                 * @return {?}
                 */
                (data) => {
                    if (data) {
                        this.libOrgGroup.getGroupList({ saveGroup: true });
                    }
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                (error) => {
                    throw new Error(this.localeService.getValue("organizationSelector.errorMessage.saveOrganization"));
                })));
            }
        })), tap((/**
         * @return {?}
         */
        () => {
        }), (/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            this.messageService.show('error', this.localeService.getValue("organizationSelector.errorMessage.saveGroup"), true);
        }))).subscribe();
    }
    /**
     * @private
     * @param {?} param
     * @return {?}
     */
    updateNameOrNote(param) {
        /** @type {?} */
        const updateInfo$ = this.apiService.saveGroup(param);
        updateInfo$.pipe(tap((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            if (data) {
                this.editingGroupId = null;
                this.groupTitle = null;
                this.groupNote = null;
                this.groupDialogTitle = this.localeService.getValue("organizationSelector.group.saveGroup");
                this.libOrgGroup.getGroupList({ updateNameOrNote: true });
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            this.messageService.show('error', this.localeService.getValue("organizationSelector.errorMessage.editGroup"), true);
        }))).subscribe();
    }
    // 点击选组织弹窗的确定或者取消按钮时触发，1为确定，0为取消
    /**
     * @param {?} type
     * @return {?}
     */
    returnResult(type) {
        /** @type {?} */
        let finalText = '';
        if (!!type) {
            this.confirmToken = true;
            this.assembleFinalText();
            finalText = this.finalText.join();
            if (this.selections && this.selections.length) {
                this.selectionsChange.emit({
                    selections: this.selections
                });
                this.onModelChange(finalText);
                this.onModelTouched(finalText);
            }
            else {
                this.onModelChange(finalText);
                this.onModelTouched(finalText);
                this.selectionsChange.emit({
                    selections: this.selections
                });
            }
        }
        else {
            if (this.mapFields) {
                this.mapFields2Component();
            }
            else {
                this.assembleDataByOrgIds(this.orgIds);
            }
        }
        // this.orgIds = [];
        this.dialog.close();
        this.isShow = false;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    changeTabs(e) {
        if (!this.hideSearchPanel) {
            this.hideSearchPanel = true;
            this.onSearchToken = false;
        }
        this.currentTabId = e.nextId;
        if (this.showTabIds.includes(this.organizationTabId)) {
            this.libOrgTree.currentTabId = this.currentTabId;
        }
        if (e.nextId === this.groupTabId) {
            if (!this.libOrgGroup.groupData.public.length && !this.libOrgGroup.groupData.personal.length) {
                this.libOrgGroup.assembleData(true);
            }
        }
        else if (e.nextId === this.organizationTabId) {
            if (this.libOrgTree.initLoad) {
                this.libOrgTree.loadTreeData();
            }
            else {
                this.libOrgTree.farrisTree.clearAll();
                this.libOrgTree.farrisTree.rowHeight = 28;
                if (this.multiSelect) {
                    this.libOrgTree.checkOrgTreeNodes();
                }
                else {
                    this.libOrgTree.selectOrgTreeNode();
                }
            }
        }
    }
}
OrganizationSelectorComponent.decorators = [
    { type: Component, args: [{
                selector: 'farris-organization-selector',
                template: "<div class=\"org-select-help\" *ngIf=\"showInput\">\r\n  <input-group style=\"flex: 1\" *ngIf=\"!readonly && viewType === 'text'\" [placeholder]=\"'organizationSelector.selection.placeHolder' | locale\"\r\n    [groupText]=\"lookupIcon\" [readonly]=\"false\" [editable]=\"false\" [(ngModel)]=\"finalText\"\r\n    (clickHandle)=\"showFarrisDialog()\" (clear)=\"clearAll()\">\r\n  </input-group>\r\n  <div #tagbox class=\"input-group s-input-group\" *ngIf=\"viewType == 'tag'\" [class.farris-input-group-readonly]=\"readonly\">\r\n    <div class=\"form-control f-cmp-inputgroup--multi-wrapper multi-more\">\r\n      <div class=\"multi--content\" *ngFor=\"let item of finalText; index as i\" [title]=\"item\">\r\n        <span class=\"multi--item\">\r\n          {{item}}\r\n        </span>\r\n        <i class=\"f-icon f-icon-close cursor-pointer\" *ngIf=\"!readonly\" style=\"margin-top: 1px;\"\r\n          (click)=\"removeItem(item, i)\"></i>\r\n      </div>\r\n      <span class=\"multi-another\" *ngIf=\"finalText.length === 0\">{{'organizationSelector.selection.placeHolder' | locale}}</span>\r\n    </div>\r\n    <div *ngIf=\"!readonly\" class=\"input-group-append\" title=\"\">\r\n      <span *ngIf=\"finalText && finalText.length > 0\" class=\"input-group-text input-group-clear cursor-pointer\"\r\n        (click)=\"clearAll()\">\r\n        <i class=\"f-icon modal_close\"></i>\r\n      </span>\r\n      <span class=\"input-group-text cursor-pointer\" (click)=\"showFarrisDialog()\">\r\n        <span class=\"f-icon f-icon-lookup\"></span>\r\n      </span>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n\r\n<farris-dialog #farrisDialog *ngIf=\"isShow\" class=\"fos-dialog\" [resizable]=\"false\" id=\"{{dialogId}}\"\r\n  [draggable]=\"draggable\" [buttons]=\"buttonRef\" [beforeClose]=\"farrisDialogbeforeClose\" [showMaxButton]=\"false\"\r\n  [width]=\"760\" [height]=\"dialogHeight\" [title]=\"'organizationSelector.selection.selectionDialogTitle' | locale\">\r\n  <div class=\"fos-content\">\r\n    <div class=\"fos-content-container\">\r\n      <div class=\"fos-content-main\">\r\n        <lib-search #searchResultPage [primaryField]=\"primaryField\" [multiSelect]=\"multiSelect\"\r\n          [enableFavorites]=\"enableFavorites\" [absOrgType]=\"absOrgType\" [notGetStopOrg]=\"notGetStopOrg\"\r\n          [onSearch]=\"onSearchToken\" [searchText]=\"searchText\" [pageSize]=\"searchPageSize\"\r\n          [filterType]=\"filterType\" [disableOrgList]=\"disableOrgList\"\r\n          (closePanelChange)=\"listenClosePanelChange($event)\" [hidden]=\"hideSearchPanel\">\r\n        </lib-search>\r\n        <input-group class=\"fos-search-input\" id=\"org-search-input\" [placeholder]=\"'organizationSelector.selection.searchPlaceHolder' | locale\"\r\n          [(value)]=\"searchText\" [groupText]=\"searchIcon\" (enterHandle)=\"onSearch()\" (clickHandle)=\"onSearch()\"\r\n          (keyup)=\"listenSearchText($event)\" (clear)=\"clearSearch()\" *ngIf=\"!dataPermission\">\r\n        </input-group>\r\n        <farris-tabs #selectTabs class=\"fos-tabs\" [value]=\"activeTabId\" [width]=\"'100%'\" [contentFill]=\"true\" (tabChange)=\"changeTabs($event)\">\r\n          <!-- \u7EC4\u7EC7\u673A\u6784  START -->\r\n          <farris-tab [title]=\"'organizationSelector.selection.organization' | locale\" [id]=\"organizationTabId\" *ngIf=\"showTabIds.indexOf(organizationTabId) !== -1\">\r\n            <ng-template #content>\r\n              <lib-org #libOrgTree [enableHierarchicalLoading]=\"enableHierarchicalLoading\" [notGetStopOrg]=\"notGetStopOrg\"\r\n                [enableFavorites]=\"enableFavorites\" [primaryField]=\"primaryField\" [cascadeCheck]=\"cascadeCheck\"\r\n                [cascadeDown]=\"cascadeDown\" [cascadeUp]=\"cascadeUp\" [keepSelect]=\"keepSelect\" [virtualizedLoad]=\"virtualizedLoad\"\r\n                [multiSelect]=\"multiSelect\" [absOrgType]=\"absOrgType\" [layer]=\"layer\" [orgIds]=\"orgIds\"\r\n                [selectedLimit]=\"selectedLimit\" [showCheckAll]=\"showCheckAll\" [activeTabId]=\"activeTabId\" [isActiveTab]=\"orgTabActive\"\r\n                [dataPermission]=\"dataPermission\" [dataPermissionIds]=\"dataPermissionIds\" [filterType]=\"filterType\" [disableOrgList]=\"disableOrgList\">\r\n              </lib-org>\r\n            </ng-template>\r\n          </farris-tab>\r\n          <!-- \u7EC4\u7EC7\u673A\u6784  END -->\r\n          <!-- \u6211\u7684\u6536\u85CF  START -->\r\n          <farris-tab [title]=\"'organizationSelector.selection.favorites' | locale\" [id]=\"favoritesTabId\" *ngIf=\"showTabIds.indexOf(favoritesTabId) !== -1\">\r\n            <ng-template #content>\r\n              <lib-favorites #libOrgFavorites [primaryField]=\"primaryField\" [multiSelect]=\"multiSelect\"\r\n                [filterType]=\"filterType\" [disableOrgList]=\"disableOrgList\">\r\n              </lib-favorites>\r\n            </ng-template>\r\n          </farris-tab>\r\n          <!-- \u6211\u7684\u6536\u85CF  END -->\r\n          <!-- \u5E38\u7528\u7FA4\u7EC4\u5C55\u793A\u533A  START -->\r\n          <farris-tab [title]=\"'organizationSelector.selection.group' | locale\" [id]=\"groupTabId\" *ngIf=\"showTabIds.indexOf(groupTabId) !== -1\">\r\n            <ng-template #content>\r\n              <lib-group #libOrgGroup \r\n                [primaryField]=\"primaryField\" [absOrgType]=\"absOrgType\" [multiSelect]=\"multiSelect\"\r\n                [enableFavorites]=\"enableFavorites\" [groupApiUrl]=\"groupApiUrl\" [notGetStopOrg]=\"notGetStopOrg\"\r\n                [enableHierarchicalLoading]=\"enableHierarchicalLoading\" [cascadeCheck]=\"cascadeCheck\"\r\n                [cascadeDown]=\"cascadeDown\" [cascadeUp]=\"cascadeUp\" [keepSelect]=\"keepSelect\" [virtualizedLoad]=\"virtualizedLoad\"\r\n                [layer]=\"layer\" [selectedLimit]=\"selectedLimit\" [showCheckAll]=\"showCheckAll\" [isActiveTab]=\"groupTabActive\"\r\n                [filterType]=\"filterType\" [disableOrgList]=\"disableOrgList\" (groupChange)=\"listenGroupChange($event)\">\r\n              </lib-group>\r\n            </ng-template>\r\n          </farris-tab>\r\n          <!-- \u5E38\u7528\u7FA4\u7EC4\u5C55\u793A\u533A  END -->\r\n        </farris-tabs>\r\n      </div>\r\n    </div>\r\n\r\n  </div>\r\n  <div class=\"s-personnel-group\" *ngIf=\"groupDialogOpend\">\r\n    <div class=\"s-personnel-group-title\">\r\n      <span class=\"f-icon f-sidebar-close\" (click)=\"closeGroupDialog()\"></span>{{groupDialogTitle}}\r\n    </div>\r\n    <div class=\"f-form-layout farris-form farris-form-controls-inline\" style=\"margin-top: 22px;\">\r\n      <div class=\"col-12 col-md-12 col-xl-12 col-el-12\">\r\n        <div class=\"farris-group-wrap\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\" style=\"width: 4.2rem\" for=\"month_4c7f0115_7796_42b4_bf41_beb7299dd84f_caum\">\r\n              <span class=\"farris-label-info text-danger\">*</span>\r\n              <span class=\"farris-label-text\">{{'organizationSelector.group.newGroupName' | locale}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n                <input class=\"form-control\" id=\"month_4c7f0115_7796_42b4_bf41_beb7299dd84f_caum\"\r\n                (keyup)=\"listenGroupTitle($event)\" placeholder=\"{{'organizationSelector.group.newGroupPlaceHolder' | locale}}\"\r\n                [ngStyle]=\"{'border-color': !validGroupTitle ? 'red' : '#D8DCE6'}\" [(ngModel)]=\"groupTitle\" />\r\n              <!--\r\n                <language-textbox id=\"month_4c7f0115_7796_42b4_bf41_beb7299dd84f_caum\" \r\n                [ngStyle]=\"{'border-color': !validGroupTitle ? 'red' : '#D8DCE6'}\"\r\n                [languages]=\"languagesArray\" [editable]=\"true\" [enableClear]=\"true\" [(ngModel)]=\"groupTitle\"\r\n                (keyup)=\"listenGroupTitle($event)\">\r\n              </language-textbox>\r\n              -->\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <div class=\"col-12 col-md-12 col-xl-12 col-el-12\" style=\"margin-top: 4px;\">\r\n        <div class=\"farris-group-wrap\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\" style=\"width: 4.2rem\"\r\n              for=\"billNote_7b43d269_2dd3_4777_8cfb_bd0c721575cf_hh9s\">\r\n              <span class=\"farris-label-text\">{{'organizationSelector.group.newGroupNote' | locale}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <textarea class=\"form-control\" style=\"height: 66px;\" [(ngModel)]=\"groupNote\"\r\n                id=\"billNote_7b43d269_2dd3_4777_8cfb_bd0c721575cf_hh9s\" placeholder=\"{{'organizationSelector.group.newGroupPlaceHolder' | locale}}\"></textarea>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n    <div class=\"s-personnel-group-footer\">\r\n      <input type=\"button\" class=\"btn btn-primary fr\" style=\"width: 100px;\" value=\"{{'organizationSelector.selection.confirm' | locale}}\" (click)=\"saveGroup()\" />\r\n      <input type=\"button\" class=\"btn btn-secondary fr mr-3\" value=\"{{'organizationSelector.selection.cancel' | locale}}\" (click)=\"closeGroupDialog()\" />\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n\r\n<ng-template #buttonRef>\r\n  <div class=\"fos-dialog-footer\">\r\n    <span class=\"fosd-footer-selection\" (mouseover)=\"selectionsShow=true\"\r\n      (mouseout)=\"selectionsShow=false\">{{'organizationSelector.selection.firstSelectionText' | locale}}\r\n      <a>{{selections.length}}</a>{{'organizationSelector.selection.lastSelectionText' | locale}}</span>\r\n    <span class=\"fosd-footer-clear\" (click)=\"clearSeletions()\">{{'organizationSelector.selection.clear' | locale}}</span>\r\n    <lib-selection #selectedComponent class=\"fosd-selection-content\" [primaryField]=\"primaryField\"\r\n      [enableGroup]=\"enableGroup\" (mouseover)=\"selectionsShow=true\" (mouseout)=\"selectionsShow=false\"\r\n      [hidden]=\"!selectionsShow\" (openedChange)=\"listenOpenGroup($event)\"></lib-selection>\r\n  </div>\r\n  <input type=\"button\" class=\"btn btn-secondary\" value=\"{{'organizationSelector.selection.cancel' | locale}}\" (click)=\"returnResult(0)\" />\r\n  <input type=\"button\" class=\"btn btn-primary\" style=\"width: 100px;\" value=\"{{'organizationSelector.selection.confirm' | locale}}\" (click)=\"returnResult(1)\" />\r\n</ng-template>",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => OrganizationSelectorComponent)),
                        multi: true
                    },
                    FavoritesTool, ApiService, StateMgrTool
                ],
                styles: [".fos-dialog .fos-content{width:100%;height:100%;min-width:760px;overflow:hidden}.fos-dialog ::ng-deep .modal-content{border-radius:10px}.fos-dialog ::ng-deep .modal-header{color:#2d2f33;border-bottom:none;padding:15px 0 15px 24px}.fos-dialog ::ng-deep .actions{padding-right:22px;font-size:20px}.fos-dialog ::ng-deep .modal-title-label{font-size:17px;font-weight:600}.fos-dialog ::ng-deep .modal-footer{padding:17px 24px;background:#fff;border-top:none}.fos-dialog .fos-content .fos-content-container{display:-webkit-box;display:flex;width:calc(100% - 48px);height:100%;margin:0 24px;border:1px solid #dee1ea;border-radius:3px}.fos-dialog .fos-content .fos-content-container .fos-content-main{-webkit-box-flex:0;flex:0 0 100%;position:relative;display:-webkit-box;display:flex}.fos-dialog .fos-content .fos-content-container .fos-content-main .fos-search-input{position:absolute;top:7px;right:14px;color:#959ab1;z-index:99;width:320px}.fos-dialog .fos-content .fos-content-container .fos-content-main .fos-tabs{min-width:575px}.fos-dialog ::ng-deep .fos-tabs .nav-link{line-height:18px;padding:11px 0!important}.fos-dialog ::ng-deep .fos-tabs .farris-tabs-header{background:#f9f9f9}.fos-dialog ::ng-deep .fos-tabs .nav-item{margin-right:30px}.fos-dialog ::ng-deep .fos-tabs .nav{padding-left:14px;font-size:16px;font-weight:600}.fos-dialog ::ng-deep .fos-tabs .nav-item .st-tab-text{color:#6f8cb4}.fos-dialog ::ng-deep .fos-tabs .nav-item .active .st-tab-text{color:#2a87ff}.fos-dialog::ng-deep input::-webkit-input-placeholder{color:#b4bccc;font-size:13px;line-height:24px}.fos-dialog ::ng-deep input::-moz-input-placeholder{color:#b4bccc;font-size:13px;line-height:24px}.fos-dialog ::ng-deep input::-ms-input-placeholder{color:#b4bccc;font-size:13px;line-height:24px}.fos-dialog ::ng-deep .fos-search-input .form-control{line-height:26px;height:26px;font-size:13px}.fos-dialog ::ng-deep .fos-treetable .farris-treetable-thead{background:#fff;font-size:13px}.fos-dialog ::ng-deep .fos-treetable .farris-treetable-thead th .tt-checkbox .custom-control-label::before{top:2px;margin:0}.fos-dialog ::ng-deep .fos-treetable .farris-treetable-thead th:hover{background:#fff}.fos-dialog ::ng-deep .fos-treetable .treetable-cell-content span{font-size:13px}:host .org-select-help{display:-webkit-box;display:flex;width:100%}:host .input-group .input-group-append{background:#fff}:host .s-input-group .input-group-text{color:rgba(0,0,0,.25)}:host .cursor-pointer{cursor:pointer}:host .s-input-group .multi-more{padding:0 0 2px 8px;height:auto;min-height:24px;-webkit-box-flex:1;flex:1;overflow:hidden}:host .s-input-group .multi--content{display:-webkit-box;display:flex;flex-wrap:nowrap;cursor:default;padding:0 5px;background:#ebf7fe;border:1px solid #cfedff;margin:3px 3px 0 0;float:left;height:17px;border-radius:3px;line-height:17px;-webkit-box-pack:justify;justify-content:space-between;box-sizing:content-box;max-width:calc(100% - 10px)}:host .s-input-group .multi--item{line-height:16px;text-overflow:ellipsis;height:16px;-webkit-box-flex:1;flex:1;overflow:hidden;word-break:break-all}:host .s-input-group .multi-another{padding:3px 3px 0 0;line-height:16px;float:left;height:17px;box-sizing:content-box;border:1px solid #fff;color:#bfbfbf}.fos-dialog ::ng-deep .fos-dialog-footer{width:200px;height:30px;line-height:30px;position:absolute;left:24px;text-align:left;cursor:default}.fos-dialog ::ng-deep .fos-dialog-footer .fosd-footer-selection{font-size:14px;color:#2d2f33;background:#f4f9ff;border-radius:5px;padding:2px 5px}.fos-dialog ::ng-deep .fos-dialog-footer .fosd-footer-selection a{font-size:17px;color:#2a87ff}.fos-dialog ::ng-deep .fos-dialog-footer .fosd-footer-clear{font-size:13px;color:#2a87ff;margin-left:1rem;cursor:pointer}.fos-dialog ::ng-deep .fos-dialog-footer .fosd-selection-content{position:absolute;bottom:25px;left:-10px;background:#fff;z-index:102}.fos-dialog::ng-deep .pagination-container{height:1.5rem;margin-bottom:21px}.fos-dialog::ng-deep .pagination-container input::-webkit-inner-spin-button,.fos-dialog::ng-deep .pagination-container input::-webkit-outer-spin-button{-webkit-appearance:none}.fos-dialog::ng-deep .pagination-container input[type=number]{-moz-appearance:textfield;text-align:center;width:38px;line-height:20px;border:1px solid #e9ecf3;border-radius:6px;box-shadow:0 2px 6px 0 rgba(31,35,41,.06);margin-left:4px;outline:0}.fos-dialog::ng-deep .pagination-container input[type=number]:focus{border:1px solid #0084ff}.fos-dialog::ng-deep .pagination-container li{line-height:22px}.fos-dialog ::ng-deep .fos-treetable td{border-color:#fff}.fos-dialog ::ng-deep .default-light-color{color:#f7b500}.fos-dialog ::ng-deep .default-dark-color{color:#d8dce6}.fos-dialog ::ng-deep .s-personnel-group{width:440px;height:250px;border-radius:5px;background:#fff;box-shadow:0 0 40px 0 rgba(0,0,0,.2);padding:16px 24px;position:fixed;top:50%;left:50%;margin-left:-220px;margin-top:-125px;z-index:9999}.fos-dialog ::ng-deep .s-personnel-group .s-personnel-group-title{font-size:17px;line-height:24px;color:#2d2f33;font-weight:600}.fos-dialog ::ng-deep .s-personnel-group .s-personnel-group-title .f-icon{float:right;font-size:20px;color:#878d99;margin-top:3px;cursor:pointer}.fos-dialog ::ng-deep .s-personnel-group .s-personnel-group-footer{margin-top:20px}.fos-dialog ::ng-deep .s-personnel-group .col-12{padding:0}.fos-dialog ::ng-deep .s-personnel-group-footer .title-checkall{margin-top:6px}.fos-dialog ::ng-deep .s-personnel-group .farris-label-text{color:#5a5e66}.fos-dialog ::ng-deep .s-personnel-group .s-personnel-group-footer .fr{float:right;height:32px}"]
            }] }
];
/** @nocollapse */
OrganizationSelectorComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: Injector },
    { type: FavoritesTool },
    { type: ApiService },
    { type: NgZone },
    { type: StateMgrTool }
];
OrganizationSelectorComponent.propDecorators = {
    baseCls: [{ type: HostBinding, args: ['class.f-organization-selector',] }],
    width: [{ type: HostBinding, args: ['style.width',] }, { type: Input }],
    dialog: [{ type: ViewChild, args: ['farrisDialog',] }],
    libOrgGroup: [{ type: ViewChild, args: ['libOrgGroup',] }],
    libOrgTree: [{ type: ViewChild, args: ['libOrgTree',] }],
    libOrgFavorites: [{ type: ViewChild, args: ['libOrgFavorites',] }],
    showInput: [{ type: Input }],
    busNum: [{ type: Input }],
    multiSelect: [{ type: Input }],
    placeholder: [{ type: Input }],
    searchPlaceholder: [{ type: Input }],
    readonly: [{ type: Input }],
    absUserOrgUrl: [{ type: Input }],
    absRfQueryUrl: [{ type: Input }],
    absRfModifyUrl: [{ type: Input }],
    groupApiUrl: [{ type: Input }],
    groupExtendUrl: [{ type: Input }],
    absOrgType: [{ type: Input }],
    enableHierarchicalLoading: [{ type: Input }],
    notGetStopOrg: [{ type: Input }],
    primaryField: [{ type: Input }],
    displayField: [{ type: Input }],
    orgIds: [{ type: Input }],
    layer: [{ type: Input }],
    cascadeCheck: [{ type: Input }],
    cascadeDown: [{ type: Input }],
    cascadeUp: [{ type: Input }],
    showCheckAll: [{ type: Input }],
    keepSelect: [{ type: Input }],
    formatFn: [{ type: Input }],
    selectedLimit: [{ type: Input }],
    showTabIds: [{ type: Input }],
    activeTabId: [{ type: Input }],
    searchPageSize: [{ type: Input }],
    mapFields: [{ type: Input }],
    multipleChoiceSeparator: [{ type: Input }],
    draggable: [{ type: Input }],
    virtualizedLoad: [{ type: Input }],
    dataPermission: [{ type: Input }],
    dataPermissionIds: [{ type: Input }],
    filterType: [{ type: Input }],
    disableOrgList: [{ type: Input }],
    filterOrgList: [{ type: Input }],
    selectionsChange: [{ type: Output }],
    inputClear: [{ type: Output }],
    tagRemoved: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    OrganizationSelectorComponent.prototype.baseCls;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.width;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.dialog;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.libOrgGroup;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.libOrgTree;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.libOrgFavorites;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.showInput;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.busNum;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.multiSelect;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.placeholder;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.searchPlaceholder;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.readonly;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.absUserOrgUrl;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.absRfQueryUrl;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.absRfModifyUrl;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.groupApiUrl;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.groupExtendUrl;
    /**
     * 数据源类型
     * 默认为： System_organization  系统数据源   获取系统人员、系统组织
     * Administrative_organization  已扩展数据源  获取行政组织、行政人员
     * absOrgType {string}
     * PersonnelSelectorComponent
     * @type {?}
     */
    OrganizationSelectorComponent.prototype.absOrgType;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.enableHierarchicalLoading;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.notGetStopOrg;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.primaryField;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.displayField;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.orgIds;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.layer;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.cascadeCheck;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.cascadeDown;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.cascadeUp;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.showCheckAll;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.keepSelect;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.formatFn;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.selectedLimit;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.showTabIds;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.activeTabId;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.searchPageSize;
    /**
     * 字段映射, 仅在farris form中使用{ 帮助数据字段：表单字段 }
     * @type {?}
     */
    OrganizationSelectorComponent.prototype.mapFields;
    /**
     * 多选分隔符
     * @type {?}
     */
    OrganizationSelectorComponent.prototype.multipleChoiceSeparator;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.draggable;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.virtualizedLoad;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.dataPermission;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.dataPermissionIds;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.filterType;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.selectionsChange;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.inputClear;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.tagRemoved;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.viewType;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.selections;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.originalSelections;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.finalText;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.lookupIcon;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.searchIcon;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.organizationTabId;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.groupTabId;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.favoritesTabId;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.dialogHeight;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.dialogId;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.isShow;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.selectionsSub;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.confirmToken;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.hideSearchPanel;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.searchText;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.onSearchToken;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.enableFavorites;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.enableGroup;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.groupDialogOpend;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.groupDialogTitle;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.groupTitle;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.groupNote;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.validGroupTitle;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.groupTitleStream;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.ngControl;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.selectionsShow;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.editingGroupId;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.cdRef;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.currentTabId;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.notifyService;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.orgTabActive;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.groupTabActive;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.againstSingleSelect;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.localeService;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.languagesArray;
    /** @type {?} */
    OrganizationSelectorComponent.prototype._disableOrgList;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.farrisDialogbeforeClose;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.onModelChange;
    /** @type {?} */
    OrganizationSelectorComponent.prototype.onModelTouched;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.render;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.favoritesTool;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.apiService;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    OrganizationSelectorComponent.prototype.stateMgr;
}
