// @gsp-aif/runtime-api/fesm5/gsp-aif-runtime-api.js
import { WindowModule, DialogModule } from '@progress/kendo-angular-dialog';
import { TreeViewModule } from '@progress/kendo-angular-treeview';
import { GridModule } from '@progress/kendo-angular-grid';
import { process } from '@progress/kendo-data-query';
import { __values } from 'tslib';
import { Subject } from 'rxjs';
import { Injectable, Injector, ComponentFactoryResolver, Component, Input, Output, ViewChild, EventEmitter, ViewEncapsulation, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { ButtonsModule } from '@progress/kendo-angular-buttons';
import { DropDownsModule } from '@progress/kendo-angular-dropdowns';
import { SplitterModule } from '@progress/kendo-angular-layout';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { HttpService, SessionService } from '@ecp-caf/caf-common';
import { BsModalService, ModalModule } from '@farris/ui-modal';
import { FarrisDialogModule } from '@farris/ui-dialog';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 生单关联实体类
 */
var  /**
 * 生单关联实体类
 */
AifAssociationEntity = /** @class */ (function () {
    function AifAssociationEntity() {
    }
    return AifAssociationEntity;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 查询生单规则请求体参数类
 */
var  /**
 * 查询生单规则请求体参数类
 */
GetCreationRuleArgs = /** @class */ (function () {
    function GetCreationRuleArgs() {
    }
    return GetCreationRuleArgs;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 单据追踪组件配置
 */
var  /**
 * 单据追踪组件配置
 */
AifDocumentsTrackManagerModuleConfig = /** @class */ (function () {
    function AifDocumentsTrackManagerModuleConfig() {
    }
    return AifDocumentsTrackManagerModuleConfig;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 拉式生单生单规则选择组件配置
 */
var  /**
 * 拉式生单生单规则选择组件配置
 */
AifPullSelectCreationRuleModuleConfig = /** @class */ (function () {
    function AifPullSelectCreationRuleModuleConfig() {
    }
    return AifPullSelectCreationRuleModuleConfig;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 推式生单生单规则选择组件配置
 */
var  /**
 * 推式生单生单规则选择组件配置
 */
AifPushSelectCreationRuleModuleConfig = /** @class */ (function () {
    function AifPushSelectCreationRuleModuleConfig() {
    }
    return AifPushSelectCreationRuleModuleConfig;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var TrackType = {
    UpSearch: "UpSearch",
    DownSearch: "DownSearch",
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 供内部和外部共同使用的单据追踪服务类，可输入参数自动查询并显示单据追踪模态框，也可以仅获取单据追踪查询结果
 */
var AifDocumentsTrackService = /** @class */ (function () {
    function AifDocumentsTrackService(http, modalService, injector, cfr, config) {
        this.http = http;
        this.modalService = modalService;
        this.injector = injector;
        this.cfr = cfr;
        this.serverIP = "";
        this.baseUrl = '/api/runtime/aif/v1.0/associationcontroller/';
        this.serverIP = config == null || config.serverIP == null ? "" : config.serverIP;
    }
    /**
     * 根据参数追踪并显示追踪模态框，注意：调用该方法时应同时引用TrackType枚举类
     * @param documentId 单据ID
     * @param beType 单据类型编号
     * @param trackType 追踪类型，仅可用“UpSearch”（上查）或“DownSearch”（下查)
     */
    /**
     * 根据参数追踪并显示追踪模态框，注意：调用该方法时应同时引用TrackType枚举类
     * @param {?} documentId 单据ID
     * @param {?} beType 单据类型编号
     * @param {?} trackType 追踪类型，仅可用“UpSearch”（上查）或“DownSearch”（下查)
     * @return {?}
     */
    AifDocumentsTrackService.prototype.showResultDialog = /**
     * 根据参数追踪并显示追踪模态框，注意：调用该方法时应同时引用TrackType枚举类
     * @param {?} documentId 单据ID
     * @param {?} beType 单据类型编号
     * @param {?} trackType 追踪类型，仅可用“UpSearch”（上查）或“DownSearch”（下查)
     * @return {?}
     */
    function (documentId, beType, trackType) {
        /** @type {?} */
        var cmpF = this.cfr.resolveComponentFactory(AifDocumentsTrackManagerComponent);
        /** @type {?} */
        var cmpR = cmpF.create(this.injector);
        if (trackType != TrackType.DownSearch && trackType != TrackType.UpSearch) {
            trackType = TrackType.DownSearch;
        }
        cmpR.instance.searchFromOut(documentId, beType, trackType);
        return this.modalService.show(cmpR, {
            title: '追踪',
            width: 1100,
            height: 600,
            buttons: null,
            showButtons: false
        });
    };
    /**
     * 根据参数追踪并将查询结果返回，注意：调用该方法时应同时引用TrackType枚举类
     * @param documentId 单据ID
     * @param beType 单据类型编号
     * @param trackType 追踪类型，仅可用“UpSearch”（上查）或“DownSearch”（下查),
     */
    /**
     * 根据参数追踪并将查询结果返回，注意：调用该方法时应同时引用TrackType枚举类
     * @param {?} documentId 单据ID
     * @param {?} beType 单据类型编号
     * @param {?} trackType 追踪类型，仅可用“UpSearch”（上查）或“DownSearch”（下查),
     * @return {?}
     */
    AifDocumentsTrackService.prototype.getTrackedDocumentData = /**
     * 根据参数追踪并将查询结果返回，注意：调用该方法时应同时引用TrackType枚举类
     * @param {?} documentId 单据ID
     * @param {?} beType 单据类型编号
     * @param {?} trackType 追踪类型，仅可用“UpSearch”（上查）或“DownSearch”（下查),
     * @return {?}
     */
    function (documentId, beType, trackType) {
        if (trackType == TrackType.DownSearch) {
            return this.getTargetData(beType, documentId);
        }
        else {
            return this.getSourceData(beType, documentId);
        }
    };
    /**
     * @private
     * @param {?} srcBeType
     * @param {?} srcDataId
     * @return {?}
     */
    AifDocumentsTrackService.prototype.getTargetData = /**
     * @private
     * @param {?} srcBeType
     * @param {?} srcDataId
     * @return {?}
     */
    function (srcBeType, srcDataId) {
        /** @type {?} */
        var url = this.serverIP + this.baseUrl + ("GetTargetData/" + srcBeType + "&" + srcDataId);
        return this.http.get(url);
    };
    /**
     * @private
     * @param {?} tarBeType
     * @param {?} tarDataId
     * @return {?}
     */
    AifDocumentsTrackService.prototype.getSourceData = /**
     * @private
     * @param {?} tarBeType
     * @param {?} tarDataId
     * @return {?}
     */
    function (tarBeType, tarDataId) {
        /** @type {?} */
        var url = this.serverIP + this.baseUrl + ("GetSourceData/" + tarBeType + "&" + tarDataId);
        return this.http.get(url);
    };
    AifDocumentsTrackService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AifDocumentsTrackService.ctorParameters = function () { return [
        { type: HttpService },
        { type: BsModalService },
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: AifDocumentsTrackManagerModuleConfig }
    ]; };
    return AifDocumentsTrackService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 单据追踪类型的文本映射，UpSearch为上查，DownSearch为下查，外部可用
 * @type {?}
 */
var TrackTypeTextMap = {
    UpSearch: "上查",
    DownSearch: "下查",
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 单据追踪主组件
 */
var AifDocumentsTrackManagerComponent = /** @class */ (function () {
    function AifDocumentsTrackManagerComponent(trackService) {
        this.trackService = trackService;
        /**
         * 子组件显示控制
         */
        this.isHidden = true;
        this.isShowable = false;
        this.isResultEmpty = null;
        this.documentId = null;
        this.beTypeCode = null;
        this.selectedTrackType = null;
        this.trackTypes = [];
        /**
         * 传递给子组件的所选择的追踪方式
         */
        this.selectedTrackTypeToChildren = "";
        /**
         * 响应结果转化为组件内传输格式
         */
        this.beTypeCodes = [];
        this.documents = {};
        this.beTypeNames = {};
        //初始化追踪方式下拉框
        for (var item in TrackType) {
            this.trackTypes.push({ value: item, text: TrackTypeTextMap[item] });
        }
        //默认选择“下查”方式
        this.selectedTrackType = this.trackTypes[1].value;
        this.selectedTrackTypeToChildren = this.selectedTrackType;
    }
    /**
     * @return {?}
     */
    AifDocumentsTrackManagerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    AifDocumentsTrackManagerComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
    };
    /**
     * @return {?}
     */
    AifDocumentsTrackManagerComponent.prototype.search = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.selectedTrackTypeToChildren = this.selectedTrackType;
        this.trackService.getTrackedDocumentData(this.documentId, this.beTypeCode, this.selectedTrackType).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            var e_1, _a;
            /** @type {?} */
            var tempDocuments = {};
            /** @type {?} */
            var tempBeTypeNames = {};
            /** @type {?} */
            var tempBeTypeCodes = [];
            if (_this.isNullOrUndefined(data) == false && data.length > 0) {
                _this.isResultEmpty = false;
                _this.isShowable = true;
                _this.isHidden = false;
                try {
                    for (var data_1 = __values(data), data_1_1 = data_1.next(); !data_1_1.done; data_1_1 = data_1.next()) {
                        var item = data_1_1.value;
                        //查询结果不为空时
                        if (_this.isNullOrUndefined(tempDocuments[item.BeTypeCode]) == false) {
                            tempDocuments[item.BeTypeCode].push(item);
                        }
                        else {
                            tempBeTypeCodes.push(item.BeTypeCode);
                            tempBeTypeNames[item.BeTypeCode] = item.BeTypeName;
                            tempDocuments[item.BeTypeCode] = [item];
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (data_1_1 && !data_1_1.done && (_a = data_1.return)) _a.call(data_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            else {
                //查询结果为空时
                _this.isResultEmpty = true;
                _this.isShowable = false;
                _this.isHidden = true;
            }
            _this.documents = tempDocuments;
            _this.beTypeCodes = tempBeTypeCodes;
            _this.beTypeNames = tempBeTypeNames;
        }));
    };
    /**
     * 为外部提供的方便递归调用的查询接口
     * @param documentId
     * @param beType
     * @param trackType
     */
    /**
     * 为外部提供的方便递归调用的查询接口
     * @param {?} documentId
     * @param {?} beType
     * @param {?} trackType
     * @return {?}
     */
    AifDocumentsTrackManagerComponent.prototype.searchFromOut = /**
     * 为外部提供的方便递归调用的查询接口
     * @param {?} documentId
     * @param {?} beType
     * @param {?} trackType
     * @return {?}
     */
    function (documentId, beType, trackType) {
        this.documentId = documentId;
        this.beTypeCode = beType;
        this.selectedTrackType = trackType;
        this.search();
    };
    /**
     * 从查询结果中选择单据继续追踪时，父组件接收到子组件传来的
     * 对象，并自动调用查询方法
     * @param $event
     */
    /**
     * 从查询结果中选择单据继续追踪时，父组件接收到子组件传来的
     * 对象，并自动调用查询方法
     * @param {?} $event
     * @return {?}
     */
    AifDocumentsTrackManagerComponent.prototype.trackGoingOn = /**
     * 从查询结果中选择单据继续追踪时，父组件接收到子组件传来的
     * 对象，并自动调用查询方法
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.documentId = $event.documentId;
        this.beTypeCode = $event.beTypeCode;
        this.selectedTrackType = $event.trackType;
        this.search();
    };
    /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    AifDocumentsTrackManagerComponent.prototype.isNullOrUndefined = /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        return arg == null;
    };
    AifDocumentsTrackManagerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-aif-documents-track-manager',
                    template: "<div class=\"d-flex g-main\">\r\n  <div class=\"col\">\r\n    <div class=\"g-main d-flex flex-column px-2\">\r\n      <!--\u4E3B\u5185\u5BB9\u533A-->\r\n      <div class=\"mb-1 pr-2 position-relative clearfix\">\r\n        <!-- <div class=\"\" style=\"\"> -->\r\n        <div class=\"position-relative\" [style.display]=\"'inherit'\">\r\n          <div class=\"form-inline row py-1\">\r\n            <div class=\"form-group col-sm-12 col-md-4 mb-1\">\r\n              <div class=\"input-group w-100\">\r\n                <div class=\"input-group-prepend\"><span class=\"input-group-text\">\u5355\u636EID</span></div>\r\n                <input class=\"form-control\" placeholder=\"\" name=\"documentId\" [(ngModel)]=\"documentId\" type=\"text\" />\r\n              </div>\r\n            </div>\r\n\r\n            <div class=\"form-group col-sm-12 col-md-4 mb-1\">\r\n                <div class=\"input-group w-100\">\r\n                  <div class=\"input-group-prepend\"><span class=\"input-group-text\">BE\u7C7B\u578B\u7F16\u53F7</span></div>\r\n                  <input class=\"form-control\" placeholder=\"\" name=\"beTypeCode\" [(ngModel)]=\"beTypeCode\" type=\"text\" />\r\n                </div>\r\n            </div>\r\n\r\n            <div class=\"form-group col-sm-12 col-md-2 mb-1\">\r\n                <div class=\"input-group w-100\">\r\n                  <div class=\"input-group-prepend\"><span class=\"input-group-text\">\u8FFD\u8E2A\u65B9\u5F0F</span></div>\r\n                  <select [(ngModel)]=\"selectedTrackType\" class=\"form-control\">\r\n                      <option *ngFor=\"let tt of trackTypes\" value=\"{{tt.value}}\">{{tt.text}}</option>\r\n                    </select>\r\n                </div>\r\n            </div>\r\n\r\n            <div class=\" col-sm-12 col-md-1 mb-1\">\r\n              <div class=\"w-100\">\r\n                <button class=\"k-button btn btn-sm btn-default\" type=\"button\" (click)=\"search()\">\u67E5\u8BE2</button>\r\n              </div>\r\n            </div>\r\n          </div>\r\n\r\n        </div>\r\n        <div class=\"k-form-field\" style=\"color:red; font-size: unset;\" *ngIf=\"isResultEmpty==true\">\u67E5\u8BE2\u7ED3\u679C\u4E3A\u7A7A</div>\r\n      </div>\r\n\r\n      <div class=\"flex-auto k-form-field pb-1\">\r\n        <app-aif-documents-tree [documents]=\"documents\" [beTypeNames]=\"beTypeNames\" [beTypeCodes]=\"beTypeCodes\"\r\n          [trackType]=\"selectedTrackTypeToChildren\" (trackClick)=\"trackGoingOn($event)\">\r\n        </app-aif-documents-tree>\r\n      </div>\r\n\r\n    </div>\r\n  </div>\r\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    AifDocumentsTrackManagerComponent.ctorParameters = function () { return [
        { type: AifDocumentsTrackService }
    ]; };
    AifDocumentsTrackManagerComponent.propDecorators = {
        documentId: [{ type: Input }],
        beTypeCode: [{ type: Input }],
        selectedTrackType: [{ type: Input }],
        trackTypes: [{ type: Output }]
    };
    return AifDocumentsTrackManagerComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 预置单据属性与显示文本的映射，外部可用并可以对该对象进行修改
 * @type {?}
 */
var AifDocumentsPropertyTextMap = {
    DataId: "ID",
    PrimaryValue: "ID",
    BeTypeCode: "单据类型编号",
    BeTypeName: "单据类型名称",
    TotalValue: "总价",
    TotalNumber: "总数量"
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 供查询结果组件使用的文本映射，上查时查询结果分类表头为上游单据，下查时表头为下游单据，不能识别时显示单据类型
 * @type {?}
 */
var TrackDirectionTypeMap = {
    UnKnown: "单据类型",
    UpSearch: "上游单据",
    DownSearch: "下游单据",
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 单据追踪查询结果组件
 */
var AifDocumentsTreeComponent = /** @class */ (function () {
    function AifDocumentsTreeComponent(trackService) {
        this.trackService = trackService;
        /**
         * 单据记录Grid
         */
        this.stateOfDocuments = {
            skip: 0,
            take: 30,
        };
        this.checkboxOnly = false;
        this.disabledState = true;
        this.selectedByFieldOfDocuments = "";
        this.selectableSettings = {
            checkboxOnly: this.checkboxOnly,
            mode: "single"
        };
        this.documentProperties = [];
        this.gridDataOfDocuments = null;
        this.selectedKeysOfDocuments = new Array();
        /**
         * 单据分类导航
         */
        this.treeDataOfBeTypeCodes = [];
        this.selectedKeysOfTreeView = ['0'];
        this.gridDataOfBeTypes = [];
        this.selectedKeysOfBeTypes = new Array();
        this.disabledStateOfBeTypes = true;
        this.selectedByFieldOfBeTypes = "BeTypeCode";
        this.selectableSettingsOfBeTypes = {
            checkboxOnly: this.checkboxOnly,
            mode: "single"
        };
        this.selectBeType = "";
        /**
         * 常量映射
         */
        this.documentsPropertyTextMap = AifDocumentsPropertyTextMap;
        this.trackDirectionTypeMap = TrackDirectionTypeMap;
        this.trackDirectionType = TrackDirectionTypeMap.UnKnown;
        /**
         * 父子组件通信传值
         */
        this.trackClick = new EventEmitter();
        this._documents = {};
        this._beTypeCodes = [];
        this.beTypeNames = {};
    }
    Object.defineProperty(AifDocumentsTreeComponent.prototype, "documents", {
        get: /**
         * @return {?}
         */
        function () {
            return this._documents;
        },
        set: /**
         * @param {?} documents
         * @return {?}
         */
        function (documents) {
            this._documents = documents;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AifDocumentsTreeComponent.prototype, "beTypeCodes", {
        get: /**
         * @return {?}
         */
        function () {
            return this._beTypeCodes;
        },
        set: /**
         * @param {?} beTypeCodes
         * @return {?}
         */
        function (beTypeCodes) {
            var e_1, _a;
            this._beTypeCodes = beTypeCodes;
            /**重置数据源 */
            this.treeDataOfBeTypeCodes = [];
            this.documentProperties = [];
            this.gridDataOfBeTypes = [];
            this.selectedKeysOfBeTypes = [];
            this.gridDataOfDocuments = process([], this.stateOfDocuments);
            if (this.isNullOrUndefined(beTypeCodes) == false && beTypeCodes.length > 0) {
                this.selectBeType = beTypeCodes[0];
                this.selectedKeysOfBeTypes.push(beTypeCodes[0]);
            }
            try {
                for (var beTypeCodes_1 = __values(beTypeCodes), beTypeCodes_1_1 = beTypeCodes_1.next(); !beTypeCodes_1_1.done; beTypeCodes_1_1 = beTypeCodes_1.next()) {
                    var item = beTypeCodes_1_1.value;
                    // this.data[0].items.push({text:item});
                    this.gridDataOfBeTypes.push({ BeTypeCode: item, BeTypeName: this.beTypeNames[item], Number: this.documents[item].length });
                    this.treeDataOfBeTypeCodes.push({ text: this.beTypeNames[item], value: item });
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (beTypeCodes_1_1 && !beTypeCodes_1_1.done && (_a = beTypeCodes_1.return)) _a.call(beTypeCodes_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            if (this.isNullOrUndefined(this.documents) == false && this.isNullOrUndefined(this.documents[this.selectBeType]) == false && this.documents[this.selectBeType].length > 0) {
                this.loadItems();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AifDocumentsTreeComponent.prototype, "trackType", {
        get: /**
         * @return {?}
         */
        function () {
            return this._trackType;
        },
        set: /**
         * @param {?} trackType
         * @return {?}
         */
        function (trackType) {
            this._trackType = trackType;
            this.trackDirectionType = this.trackDirectionTypeMap[this._trackType];
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    AifDocumentsTreeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    AifDocumentsTreeComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        console.log(changes);
        if (this.isNullOrUndefined(changes.beTypeCodes) == false) {
            this.beTypeCodes = changes.beTypeCodes.currentValue;
        }
        if (this.isNullOrUndefined(changes.documents) == false) {
            this.documents = changes.documents.currentValue;
            // this.gridDataOfDocuments=process(this.documents,this.stateOfDocuments);
        }
        if (this.isNullOrUndefined(changes.trackType) == false) {
            this.trackType = changes.trackType.currentValue;
        }
        if (this.isNullOrUndefined(changes.beTypeNames) == false) {
            this.beTypeNames = changes.beTypeNames.currentValue;
        }
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AifDocumentsTreeComponent.prototype.clickNode = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var item = obj.item.dataItem;
        if (this.isNullOrUndefined(item.items) == false) {
            return;
        }
        console.log(item.text);
        this.selectBeType = item.value;
        this.loadItems();
        this.selectedKeysOfDocuments.splice(0, this.selectedKeysOfDocuments.length);
    };
    /**
     * @return {?}
     */
    AifDocumentsTreeComponent.prototype.selectedKeysChangeOfDocuments = /**
     * @return {?}
     */
    function () {
        if (this.selectedKeysOfDocuments.length > 0) {
            this.disabledState = false;
        }
        else {
            this.disabledState = true;
        }
    };
    /**
     * @return {?}
     */
    AifDocumentsTreeComponent.prototype.selectedKeysChangeOfBeTypes = /**
     * @return {?}
     */
    function () {
        console.log(this.selectedKeysOfBeTypes);
        if (this.selectedKeysOfBeTypes.length > 0) {
            this.disabledStateOfBeTypes = false;
            this.selectBeType = this.selectedKeysOfBeTypes[0];
            //页数置为初始值0
            this.stateOfDocuments.skip = 0;
            this.loadItems();
            //清空已选中的行
            this.selectedKeysOfDocuments.splice(0, this.selectedKeysOfDocuments.length);
        }
        else {
            this.disabledStateOfBeTypes = true;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    AifDocumentsTreeComponent.prototype.pageChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.stateOfDocuments.skip = event.skip;
        this.loadItems();
    };
    /**
     * @return {?}
     */
    AifDocumentsTreeComponent.prototype.loadItems = /**
     * @return {?}
     */
    function () {
        // let temp = this.documents[this.selectBeType].slice(this.stateOfDocuments.skip, this.stateOfDocuments.skip + this.stateOfDocuments.take);
        this.documentProperties.splice(0, this.documentProperties.length);
        for (var index in this.documents[this.selectBeType][0]) {
            this.documentProperties.push(index);
        }
        this.selectedByFieldOfDocuments = this.documentProperties[0];
        this.gridDataOfDocuments = process(this.documents[this.selectBeType], this.stateOfDocuments);
        // this.fitColumns();
        // this.gridDataOfDocuments = {
        //   data: temp,
        //   total: this.documents[this.selectBeType].length
        // };
    };
    /**
     * @return {?}
     */
    AifDocumentsTreeComponent.prototype.getDocumentDetailInfo = /**
     * @return {?}
     */
    function () {
        if (this.selectedKeysOfDocuments.length == 0) {
            alert("请选择一行数据");
            return;
        }
        console.log(this.selectedKeysOfDocuments);
        alert("详细信息窗口");
    };
    /**
     * @return {?}
     */
    AifDocumentsTreeComponent.prototype.track = /**
     * @return {?}
     */
    function () {
        if (this.selectedKeysOfDocuments.length == 0) {
            alert("请选择一行数据");
            return;
        }
        this.trackService.showResultDialog(this.selectedKeysOfDocuments[0], this.selectBeType, this.trackType);
    };
    /**
     * @param {?} context
     * @return {?}
     */
    AifDocumentsTreeComponent.prototype.rowCallback = /**
     * @param {?} context
     * @return {?}
     */
    function (context) {
        /** @type {?} */
        var isEven = context.index % 2 == 0;
        return {
            even: isEven,
            odd: !isEven
        };
    };
    /**
     * @param {?} state
     * @return {?}
     */
    AifDocumentsTreeComponent.prototype.dataStateChangeOfDocuments = /**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this.stateOfDocuments = state;
        this.gridDataOfDocuments = process(this.documents[this.selectBeType], this.stateOfDocuments);
    };
    // 自动调整列宽
    // fitColumns(): void {
    //   this.ngZone.onStable.asObservable().pipe(take(1)).subscribe(() => {
    //     this.gridOfDocuments.autoFitColumns();
    //     // console.log("this.gridOfDocuments.autoFitColumns()");
    //   });
    // }
    // 自动调整列宽
    // fitColumns(): void {
    //   this.ngZone.onStable.asObservable().pipe(take(1)).subscribe(() => {
    //     this.gridOfDocuments.autoFitColumns();
    //     // console.log("this.gridOfDocuments.autoFitColumns()");
    //   });
    // }
    /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    AifDocumentsTreeComponent.prototype.isNullOrUndefined = 
    // 自动调整列宽
    // fitColumns(): void {
    //   this.ngZone.onStable.asObservable().pipe(take(1)).subscribe(() => {
    //     this.gridOfDocuments.autoFitColumns();
    //     // console.log("this.gridOfDocuments.autoFitColumns()");
    //   });
    // }
    /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        return arg == null;
    };
    AifDocumentsTreeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-aif-documents-tree',
                    encapsulation: ViewEncapsulation.None,
                    template: "<!-- <div class=\"d-flex\" style=\"padding: 5px;\"> -->\r\n<!-- <div class=\"row\"> -->\r\n<kendo-splitter orientation=\"horizontal\" style=\"height:100%;\">\r\n  <kendo-splitter-pane [collapsible]=\"true\" size=\"16%\" style=\"padding:0 1rem 1rem 1rem;\">\r\n    <div class=\"k-button-group\" style=\"height:2.5rem;\">\r\n      <label for=\"\" style=\"margin:auto 0 auto 0;\">{{trackDirectionType}}</label>\r\n    </div>\r\n    <kendo-grid [data]=\"gridDataOfBeTypes\" [resizable]=\"true\" [selectable]=\"selectableSettingsOfBeTypes\"\r\n      [kendoGridSelectBy]=\"selectedByFieldOfBeTypes\" [selectedKeys]=\"selectedKeysOfBeTypes\" (selectedKeysChange)=\"selectedKeysChangeOfBeTypes()\">\r\n      <kendo-grid-column field=\"BeTypeName\" title=\"\u540D\u79F0\"></kendo-grid-column>\r\n      <!-- <kendo-grid-column field=\"BeTypeCode\" title=\"\u7F16\u53F7\"></kendo-grid-column> -->\r\n      <kendo-grid-column field=\"Number\" title=\"\u6570\u91CF\"></kendo-grid-column>\r\n    </kendo-grid>\r\n  </kendo-splitter-pane>\r\n  <kendo-splitter-pane style=\"padding:0 1rem 2rem 1rem;\">\r\n    <div class=\"k-button-group\" style=\"height:2.5rem;padding: 0.5rem 0 0.5rem 0;\">\r\n      <button hidden class=\"k-button btn btn-sm btn-default\" (click)=\"getDocumentDetailInfo()\">\u8BE6\u60C5</button>\r\n      <button class=\"k-button btn btn-sm ml-2 btn-default\" (click)=\"track()\">\u8FFD\u8E2A</button>\r\n    </div>\r\n\r\n    <kendo-grid class=\"\" #gridOfDocuments [data]=\"gridDataOfDocuments\" [pageable]=\"true\" [resizable]=\"true\" [pageSize]=\"stateOfDocuments.take\"\r\n      (pageChange)=\"pageChange($event)\" [skip]=\"stateOfDocuments.skip\" [selectable]=\"selectableSettings\"\r\n      [kendoGridSelectBy]=\"selectedByFieldOfDocuments\" [selectedKeys]=\"selectedKeysOfDocuments\" (selectedKeysChange)=\"selectedKeysChangeOfDocuments()\"\r\n      [sortable]=\"true\" [filterable]=\"true\" [sort]=\"stateOfDocuments.sort\" [filter]=\"stateOfDocuments.filter\"\r\n      (dataStateChange)=\"dataStateChangeOfDocuments($event)\">\r\n      <!-- <ng-template kendoGridToolbarTemplate>\r\n        <button type=\"button\" class=\"k-button btn btn-sm ml-2 btn-default\" kendoGridExcelCommand icon=\"file-excel\">Export\r\n          to Excel</button>\r\n      </ng-template> -->\r\n      <!-- <ng-template ngFor let-item [ngForOf]=\"documentProperties\" let-i=\"index\">\r\n        <kendo-grid-checkbox-column *ngIf=\"beTypeCodes&&beTypeCodes.length>0\" [width]=\"40\" [headerClass]=\"{'aif-grid-header': true}\"></kendo-grid-checkbox-column>\r\n        <kendo-grid-column [headerClass]=\"{'aif-grid-header': true}\" [field]=\"item\" [title]=\"documentsPropertyTextMap[item]||item\"></kendo-grid-column>\r\n      </ng-template> -->\r\n\r\n      <kendo-grid-checkbox-column [width]=\"40\" [headerClass]=\"{'aif-grid-header': true,'aif-grid-cell-border':true}\"></kendo-grid-checkbox-column>\r\n      <kendo-grid-column *ngFor=\"let item of documentProperties\" [headerClass]=\"{'aif-grid-header': true,'aif-grid-cell-border':true}\"\r\n        [field]=\"item\" [title]=\"documentsPropertyTextMap[item]||item\"></kendo-grid-column>\r\n      <!--<kendo-grid-excel fileName=\"Products.xlsx\"></kendo-grid-excel> -->\r\n    </kendo-grid>\r\n  </kendo-splitter-pane>\r\n</kendo-splitter>\r\n\r\n<!-- </div> -->\r\n<!-- </div> -->",
                    styles: [".aif-border-green{border:1px solid green;border-radius:1rem}.aif-border-black{border:1px solid #000;border-radius:1rem}.aif-border-blue{border:1px solid #00f;border-radius:1rem}.aif-border{border:1px solid #919399;border-radius:1rem}.aif-background{background-color:#fafafa}.aif-document-tree-group{margin-top:10px;padding:1rem}.aif-margin-left{margin-left:1rem}.k-grid tr.even{background-color:#fafafa}.k-grid tr.odd{background-color:#9dc2f1}.k-grid .aif-grid-cell-border{border-right-width:1px}.aif-grid-header{background-color:#c9d7ea}.aif-min-height{height:-webkit-fill-available;height:-moz-available;height:fill-available}"]
                }] }
    ];
    /** @nocollapse */
    AifDocumentsTreeComponent.ctorParameters = function () { return [
        { type: AifDocumentsTrackService }
    ]; };
    AifDocumentsTreeComponent.propDecorators = {
        trackClick: [{ type: Output }],
        documents: [{ type: Input }],
        beTypeCodes: [{ type: Input }],
        beTypeNames: [{ type: Input }],
        trackType: [{ type: Input }],
        gridOfDocuments: [{ type: ViewChild, args: ["gridOfDocuments",] }]
    };
    return AifDocumentsTreeComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ɵ0 = {};
var AifDocumentsTrackManagerModule = /** @class */ (function () {
    function AifDocumentsTrackManagerModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    AifDocumentsTrackManagerModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            "ngModule": AifDocumentsTrackManagerModule,
            "providers": [
                { provide: AifDocumentsTrackManagerModuleConfig, useValue: config || {} }
            ]
        };
    };
    AifDocumentsTrackManagerModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        HttpClientModule,
                        FormsModule,
                        ReactiveFormsModule,
                        ButtonsModule,
                        GridModule,
                        WindowModule,
                        DropDownsModule,
                        DialogModule,
                        TreeViewModule,
                        SplitterModule,
                        GridModule,
                        ModalModule.forRoot(),
                        FarrisDialogModule,
                    ],
                    providers: [
                        SessionService,
                        HttpService,
                        { provide: AifDocumentsTrackManagerModuleConfig, useValue: ɵ0 },
                        AifDocumentsTrackService,
                    ],
                    declarations: [
                        AifDocumentsTrackManagerComponent,
                        AifDocumentsTreeComponent,
                    ],
                    bootstrap: [
                        AifDocumentsTrackManagerComponent
                    ],
                    entryComponents: [
                        AifDocumentsTrackManagerComponent
                    ],
                    exports: [
                        AifDocumentsTrackManagerComponent
                    ],
                },] }
    ];
    return AifDocumentsTrackManagerModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 模块内部服务类，也可供外部调用
 */
var AifPullSelectCreationRuleInternalService = /** @class */ (function () {
    function AifPullSelectCreationRuleInternalService(http, config) {
        this.http = http;
        this.serverIP = null;
        this.getCreationRuleUrl = "/api/runtime/aif/v1.0/creationrulecontroller/getcreationrule";
        this.serverIP = config == null || config.serverIP == null ? "" : config.serverIP;
    }
    /**
     * 访问webapi根据业务场景图ID与来源以及目标单据类型编号获取相应生单规则
     * @param args
     */
    /**
     * 访问webapi根据业务场景图ID与来源以及目标单据类型编号获取相应生单规则
     * @param {?} args
     * @return {?}
     */
    AifPullSelectCreationRuleInternalService.prototype.getCreationRule = /**
     * 访问webapi根据业务场景图ID与来源以及目标单据类型编号获取相应生单规则
     * @param {?} args
     * @return {?}
     */
    function (args) {
        return this.http.post(this.serverIP + this.getCreationRuleUrl, args);
    };
    AifPullSelectCreationRuleInternalService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AifPullSelectCreationRuleInternalService.ctorParameters = function () { return [
        { type: HttpService },
        { type: AifPullSelectCreationRuleModuleConfig }
    ]; };
    return AifPullSelectCreationRuleInternalService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var AifPullSelectCreationRuleComponent = /** @class */ (function () {
    function AifPullSelectCreationRuleComponent(crService) {
        this.crService = crService;
        this.voTypeCodes = [];
        this.voTypeObjs = {};
        this.getCreationRuleResultEntities = {};
        this.selectedVoTypeCode = null;
        this.selectedGetCreationRuleResultEntity = null;
        this.bsModalRef = null;
        this.subject = new Subject();
    }
    /**
     * @param {?} args
     * @return {?}
     */
    AifPullSelectCreationRuleComponent.prototype.getCreationRule = /**
     * @param {?} args
     * @return {?}
     */
    function (args) {
        var _this = this;
        this.crService.getCreationRule(args).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            var e_1, _a;
            try {
                for (var data_1 = __values(data), data_1_1 = data_1.next(); !data_1_1.done; data_1_1 = data_1.next()) {
                    var item = data_1_1.value;
                    if (!_this.voTypeCodes.includes(item.SourceVoType)) {
                        _this.voTypeCodes.push(item.SourceVoType);
                        _this.getCreationRuleResultEntities[item.SourceVoType] = [];
                        _this.voTypeObjs[item.SourceVoType] = {
                            "VoType": item.SourceVoType,
                            "VoName": item.SourceVoName
                        };
                    }
                    _this.getCreationRuleResultEntities[item.SourceVoType].push(item);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (data_1_1 && !data_1_1.done && (_a = data_1.return)) _a.call(data_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            if (data.length > 0) {
                _this.selectedVoTypeCode = _this.voTypeCodes[0];
                _this.selectedGetCreationRuleResultEntity = data[0];
            }
        }));
    };
    /**
     * @return {?}
     */
    AifPullSelectCreationRuleComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} voTypeCode
     * @return {?}
     */
    AifPullSelectCreationRuleComponent.prototype.changeVoTypeRadio = /**
     * @param {?} voTypeCode
     * @return {?}
     */
    function (voTypeCode) {
        this.selectedVoTypeCode = voTypeCode;
        this.selectedGetCreationRuleResultEntity = !this.isNullOrUndefined(this.getCreationRuleResultEntities[this.selectedVoTypeCode]) && this.getCreationRuleResultEntities[this.selectedVoTypeCode].length > 0 ? this.getCreationRuleResultEntities[this.selectedVoTypeCode][0] : null;
    };
    /**
     * @return {?}
     */
    AifPullSelectCreationRuleComponent.prototype.clickOk = /**
     * @return {?}
     */
    function () {
        this.subject.next((/** @type {?} */ (this.selectedGetCreationRuleResultEntity)));
        this.subject.complete();
        this.bsModalRef.close();
    };
    /**
     * @return {?}
     */
    AifPullSelectCreationRuleComponent.prototype.clickClose = /**
     * @return {?}
     */
    function () {
        this.bsModalRef.close();
    };
    /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    AifPullSelectCreationRuleComponent.prototype.isNullOrUndefined = /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        return arg == null;
    };
    AifPullSelectCreationRuleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-aif-pull-select-creation-rule',
                    template: "<div style=\"height:100%;\">\r\n    <kendo-splitter orientation=\"vertical\">\r\n        <kendo-splitter-pane [resizable]=\"false\" [scrollable]=\"false\" [collapsible]=\"false\" size=\"40%\"\r\n            class=\"aif-splitter-pane-padding\">\r\n            <div>\r\n                <div>\u9009\u62E9\u6765\u6E90\u5355\u636E\u7C7B\u578B</div>\r\n                <div style=\"height:8rem;overflow-y: auto;\">\r\n                    <label *ngFor=\"let voTypeCode of voTypeCodes;let i=index\" class=\"col-sm-12 col-md-5\"\r\n                        style=\"padding:0 0;\">\r\n                        <input type=\"radio\" name=\"voTypeCode\" class=\"\" [checked]=\"i==0\"\r\n                            [value]=\"voTypeObjs[voTypeCode].VoType\"\r\n                            (click)=\"changeVoTypeRadio(voTypeCode)\">{{voTypeObjs[voTypeCode].VoName}}\r\n                    </label>\r\n                </div>\r\n            </div>\r\n        </kendo-splitter-pane>\r\n\r\n        <kendo-splitter-pane [resizable]=\"false\" [scrollable]=\"false\" [collapsible]=\"false\" size=\"40%\"\r\n            class=\"aif-splitter-pane-padding\">\r\n            <div>\r\n                <div>\u9009\u62E9\u751F\u5355\u89C4\u5219</div>\r\n                <kendo-dropdownlist class=\"form-control\" [data]=\"getCreationRuleResultEntities[selectedVoTypeCode]\"\r\n                    textField=\"Name\" valueField=\"Id\" [(ngModel)]=\"selectedGetCreationRuleResultEntity\">\r\n                </kendo-dropdownlist>\r\n            </div>\r\n        </kendo-splitter-pane>\r\n\r\n        <kendo-splitter-pane [resizable]=\"false\" [scrollable]=\"false\" [collapsible]=\"false\" size=\"20%\"\r\n            class=\"aif-splitter-pane-padding\">\r\n            <div>\r\n                <div>\u63CF\u8FF0</div>\r\n                <textarea kendoTextArea readonly class=\"form-control\" name=\"description\" id=\"\" rows=\"3\"\r\n                    [ngModel]=\"selectedGetCreationRuleResultEntity&&selectedGetCreationRuleResultEntity.Description\"></textarea>\r\n            </div>\r\n        </kendo-splitter-pane>\r\n\r\n        <kendo-splitter-pane [resizable]=\"false\" [scrollable]=\"false\" [collapsible]=\"false\"\r\n            class=\"aif-splitter-pane-padding\">\r\n            <div style=\"float:right;\">\r\n                <button class=\"btn-primary btn btn-sm ml-2\" (click)=\"clickOk()\"\r\n                    [disabled]=\"!selectedGetCreationRuleResultEntity\">\u786E\u5B9A</button>\r\n                <button class=\"btn btn-sm ml-2 btn-default\" (click)=\"clickClose()\">\u53D6\u6D88</button>\r\n            </div>\r\n        </kendo-splitter-pane>\r\n\r\n    </kendo-splitter>\r\n</div>",
                    styles: [".aif-splitter-pane-padding{padding:.5rem 1rem}"]
                }] }
    ];
    /** @nocollapse */
    AifPullSelectCreationRuleComponent.ctorParameters = function () { return [
        { type: AifPullSelectCreationRuleInternalService }
    ]; };
    return AifPullSelectCreationRuleComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 供外部调用的拉式选择生单规则服务类接口
 */
var AifPullSelectCreationRuleService = /** @class */ (function () {
    function AifPullSelectCreationRuleService(injector, cfr, modalService) {
        this.injector = injector;
        this.cfr = cfr;
        this.modalService = modalService;
        this.innerModalOptions = {
            title: '生单规则选择',
            width: 650,
            height: 415,
            buttons: null,
            showButtons: false,
        };
    }
    /**
     * 显示选择生单规则模态框，点击确定按钮时返回选中的生单规则
     * @param args 查询生单规则所需参数
     * @param modalOptions 模态框配置,可为空
     */
    /**
     * 显示选择生单规则模态框，点击确定按钮时返回选中的生单规则
     * @param {?} args 查询生单规则所需参数
     * @param {?=} modalOptions 模态框配置,可为空
     * @return {?}
     */
    AifPullSelectCreationRuleService.prototype.showCrRuleSelectModal = /**
     * 显示选择生单规则模态框，点击确定按钮时返回选中的生单规则
     * @param {?} args 查询生单规则所需参数
     * @param {?=} modalOptions 模态框配置,可为空
     * @return {?}
     */
    function (args, modalOptions) {
        /** @type {?} */
        var cmpF = this.cfr.resolveComponentFactory(AifPullSelectCreationRuleComponent);
        /** @type {?} */
        var cmpR = cmpF.create(this.injector);
        cmpR.instance.getCreationRule(args);
        cmpR.instance.bsModalRef = this.modalService.show(cmpR, this.isNullOrUndefined(modalOptions) ? this.innerModalOptions : modalOptions);
        return cmpR.instance.subject.asObservable();
    };
    /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    AifPullSelectCreationRuleService.prototype.isNullOrUndefined = /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        return arg == null;
    };
    AifPullSelectCreationRuleService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AifPullSelectCreationRuleService.ctorParameters = function () { return [
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: BsModalService }
    ]; };
    return AifPullSelectCreationRuleService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ɵ0$1 = {};
var AifPullSelectCreationRuleModule = /** @class */ (function () {
    function AifPullSelectCreationRuleModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    AifPullSelectCreationRuleModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            "ngModule": AifPullSelectCreationRuleModule,
            "providers": [
                { provide: AifPullSelectCreationRuleModuleConfig, useValue: config || {} }
            ]
        };
    };
    AifPullSelectCreationRuleModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        HttpClientModule,
                        ButtonsModule,
                        DropDownsModule,
                        SplitterModule,
                        ModalModule,
                        FarrisDialogModule,
                    ],
                    providers: [
                        HttpService,
                        BsModalService,
                        { provide: AifPullSelectCreationRuleModuleConfig, useValue: ɵ0$1 },
                        AifPullSelectCreationRuleInternalService,
                        AifPullSelectCreationRuleService
                    ],
                    declarations: [
                        AifPullSelectCreationRuleComponent
                    ],
                    bootstrap: [
                        AifPullSelectCreationRuleComponent
                    ],
                    entryComponents: [
                        AifPullSelectCreationRuleComponent
                    ],
                    exports: []
                },] }
    ];
    return AifPullSelectCreationRuleModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 模块内部服务类，也可供外部调用
 */
var AifPushSelectCreationRuleInternalService = /** @class */ (function () {
    function AifPushSelectCreationRuleInternalService(http, config) {
        this.http = http;
        this.serverIP = "";
        this.getCreationRuleUrl = "/api/runtime/aif/v1.0/creationrulecontroller/getcreationrule";
        this.serverIP = config == null || config.serverIP == null ? "" : config.serverIP;
    }
    /**
     * 访问webapi根据业务场景图ID与源单据类型编号获取相应生单规则
     * @param bizFlowchartId 业务场景图ID
     * @param sourceVoType 源单据类型编号
     */
    /**
     * 访问webapi根据业务场景图ID与源单据类型编号获取相应生单规则
     * @param {?} args
     * @return {?}
     */
    AifPushSelectCreationRuleInternalService.prototype.getCreationRule = /**
     * 访问webapi根据业务场景图ID与源单据类型编号获取相应生单规则
     * @param {?} args
     * @return {?}
     */
    function (args) {
        return this.http.post(this.serverIP + this.getCreationRuleUrl, args);
    };
    AifPushSelectCreationRuleInternalService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AifPushSelectCreationRuleInternalService.ctorParameters = function () { return [
        { type: HttpService },
        { type: AifPushSelectCreationRuleModuleConfig }
    ]; };
    return AifPushSelectCreationRuleInternalService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var AifPushSelectCreationRuleComponent = /** @class */ (function () {
    function AifPushSelectCreationRuleComponent(crService) {
        this.crService = crService;
        this.voTypeCodes = [];
        this.voTypeObjs = {};
        this.getCreationRuleResultEntities = {};
        this.selectedVoTypeCode = null;
        this.selectedGetCreationRuleResultEntity = null;
        this.bsModalRef = null;
        this.subject = new Subject();
    }
    /**
     * @return {?}
     */
    AifPushSelectCreationRuleComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} args
     * @return {?}
     */
    AifPushSelectCreationRuleComponent.prototype.getCreationRule = /**
     * @param {?} args
     * @return {?}
     */
    function (args) {
        var _this = this;
        this.crService.getCreationRule(args).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            var e_1, _a;
            console.log(data);
            try {
                for (var data_1 = __values(data), data_1_1 = data_1.next(); !data_1_1.done; data_1_1 = data_1.next()) {
                    var item = data_1_1.value;
                    if (!_this.voTypeCodes.includes(item.TargetVoType)) {
                        _this.voTypeCodes.push(item.TargetVoType);
                        _this.getCreationRuleResultEntities[item.TargetVoType] = [];
                        _this.voTypeObjs[item.TargetVoType] = {
                            "VoType": item.TargetVoType,
                            "VoName": item.TargetVoProxyName
                        };
                    }
                    _this.getCreationRuleResultEntities[item.TargetVoType].push(item);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (data_1_1 && !data_1_1.done && (_a = data_1.return)) _a.call(data_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            if (data.length > 0) {
                _this.selectedVoTypeCode = _this.voTypeCodes[0];
                _this.selectedGetCreationRuleResultEntity = data[0];
            }
        }));
    };
    /**
     * @param {?} beTypeCode
     * @return {?}
     */
    AifPushSelectCreationRuleComponent.prototype.changeVoTypeRadio = /**
     * @param {?} beTypeCode
     * @return {?}
     */
    function (beTypeCode) {
        this.selectedVoTypeCode = beTypeCode;
        this.selectedGetCreationRuleResultEntity = !this.isNullOrUndefined(this.getCreationRuleResultEntities[this.selectedVoTypeCode]) && this.getCreationRuleResultEntities[this.selectedVoTypeCode].length > 0 ? this.getCreationRuleResultEntities[this.selectedVoTypeCode][0] : null;
    };
    /**
     * @return {?}
     */
    AifPushSelectCreationRuleComponent.prototype.clickOk = /**
     * @return {?}
     */
    function () {
        this.subject.next((/** @type {?} */ (this.selectedGetCreationRuleResultEntity)));
        this.subject.complete();
        this.bsModalRef.close();
    };
    /**
     * @return {?}
     */
    AifPushSelectCreationRuleComponent.prototype.clickClose = /**
     * @return {?}
     */
    function () {
        this.bsModalRef.close();
    };
    /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    AifPushSelectCreationRuleComponent.prototype.isNullOrUndefined = /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        return arg == null;
    };
    AifPushSelectCreationRuleComponent.decorators = [
        { type: Component, args: [{
                    selector: "app-aif-push-select-creation-rule",
                    template: "<div style=\"height:100%;\">\r\n    <kendo-splitter orientation=\"vertical\">\r\n        <kendo-splitter-pane [resizable]=\"false\" [scrollable]=\"false\" [collapsible]=\"false\" size=\"40%\"\r\n            class=\"aif-splitter-pane-padding\">\r\n            <div>\r\n                <div>\u9009\u62E9\u76EE\u6807\u5355\u636E\u7C7B\u578B</div>\r\n                <div style=\"height:8rem;overflow-y: auto;\">\r\n                    <label *ngFor=\"let voTypeCode of voTypeCodes;let i=index\" class=\"col-sm-12 col-md-5\"\r\n                        style=\"padding:0 0;\">\r\n                        <input type=\"radio\" name=\"voTypeCode\" class=\"\" [checked]=\"i==0\"\r\n                            [value]=\"voTypeObjs[voTypeCode].VoType\"\r\n                            (click)=\"changeVoTypeRadio(voTypeCode)\">{{voTypeObjs[voTypeCode].VoName}}\r\n                    </label>\r\n                </div>\r\n            </div>\r\n        </kendo-splitter-pane>\r\n\r\n        <kendo-splitter-pane [resizable]=\"false\" [scrollable]=\"false\" [collapsible]=\"false\" size=\"40%\"\r\n            class=\"aif-splitter-pane-padding\">\r\n            <div>\r\n                <div>\u9009\u62E9\u751F\u5355\u89C4\u5219</div>\r\n                <kendo-dropdownlist class=\"form-control\" [data]=\"getCreationRuleResultEntities[selectedVoTypeCode]\"\r\n                    textField=\"Name\" valueField=\"Id\" [(ngModel)]=\"selectedGetCreationRuleResultEntity\">\r\n                </kendo-dropdownlist>\r\n            </div>\r\n        </kendo-splitter-pane>\r\n\r\n        <kendo-splitter-pane [resizable]=\"false\" [scrollable]=\"false\" [collapsible]=\"false\" size=\"20%\"\r\n            class=\"aif-splitter-pane-padding\">\r\n            <div>\r\n                <div>\u63CF\u8FF0</div>\r\n                <textarea kendoTextArea readonly class=\"form-control\" name=\"description\" id=\"\" rows=\"3\"\r\n                    [ngModel]=\"selectedGetCreationRuleResultEntity&&selectedGetCreationRuleResultEntity.Description\"></textarea>\r\n            </div>\r\n        </kendo-splitter-pane>\r\n\r\n        <kendo-splitter-pane [resizable]=\"false\" [scrollable]=\"false\" [collapsible]=\"false\"\r\n            class=\"aif-splitter-pane-padding\">\r\n            <div style=\"float:right;\">\r\n                <button class=\"btn-primary btn btn-sm ml-2\" (click)=\"clickOk()\"\r\n                    [disabled]=\"!selectedGetCreationRuleResultEntity\">\u786E\u5B9A</button>\r\n                <button class=\"btn btn-sm ml-2 btn-default\" (click)=\"clickClose()\">\u53D6\u6D88</button>\r\n            </div>\r\n        </kendo-splitter-pane>\r\n\r\n    </kendo-splitter>\r\n</div>",
                    styles: [".aif-splitter-pane-padding{padding:.5rem 1rem}"]
                }] }
    ];
    /** @nocollapse */
    AifPushSelectCreationRuleComponent.ctorParameters = function () { return [
        { type: AifPushSelectCreationRuleInternalService }
    ]; };
    return AifPushSelectCreationRuleComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 供外部调用的推式选择生单规则服务类接口
 */
var AifPushSelectCreationRuleService = /** @class */ (function () {
    function AifPushSelectCreationRuleService(injector, cfr, modalService) {
        this.injector = injector;
        this.cfr = cfr;
        this.modalService = modalService;
        /**
         * 内置模态框配置
         */
        this.innerModalOptions = {
            title: '生单规则选择',
            width: 650,
            height: 415,
            buttons: null,
            showButtons: false,
        };
    }
    /**
     * 显示选择生单规则模态框，点击确定按钮时返回选中的生单规则
     * @param args 查询生单规则所需参数
     * @param modalOptions 模态框配置,可为空
     */
    /**
     * 显示选择生单规则模态框，点击确定按钮时返回选中的生单规则
     * @param {?} args 查询生单规则所需参数
     * @param {?=} modalOptions 模态框配置,可为空
     * @return {?}
     */
    AifPushSelectCreationRuleService.prototype.showCrRuleSelectModal = /**
     * 显示选择生单规则模态框，点击确定按钮时返回选中的生单规则
     * @param {?} args 查询生单规则所需参数
     * @param {?=} modalOptions 模态框配置,可为空
     * @return {?}
     */
    function (args, modalOptions) {
        /** @type {?} */
        var cmpF = this.cfr.resolveComponentFactory(AifPushSelectCreationRuleComponent);
        /** @type {?} */
        var cmpR = cmpF.create(this.injector);
        cmpR.instance.getCreationRule(args);
        cmpR.instance.bsModalRef = this.modalService.show(cmpR, this.isNullOrUndefined(modalOptions) ? this.innerModalOptions : modalOptions);
        return cmpR.instance.subject.asObservable();
    };
    /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    AifPushSelectCreationRuleService.prototype.isNullOrUndefined = /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        return arg == null;
    };
    AifPushSelectCreationRuleService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AifPushSelectCreationRuleService.ctorParameters = function () { return [
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: BsModalService }
    ]; };
    return AifPushSelectCreationRuleService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ɵ0$2 = {};
var AifPushSelectCreationRuleModule = /** @class */ (function () {
    function AifPushSelectCreationRuleModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    AifPushSelectCreationRuleModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            "ngModule": AifPushSelectCreationRuleModule,
            "providers": [
                { provide: AifPushSelectCreationRuleModuleConfig, useValue: config || {} },
            ]
        };
    };
    AifPushSelectCreationRuleModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        HttpClientModule,
                        ButtonsModule,
                        DropDownsModule,
                        SplitterModule,
                        ModalModule,
                        FarrisDialogModule,
                    ],
                    providers: [
                        HttpService,
                        BsModalService,
                        { provide: AifPushSelectCreationRuleModuleConfig, useValue: ɵ0$2 },
                        AifPushSelectCreationRuleInternalService,
                        AifPushSelectCreationRuleService
                    ],
                    declarations: [
                        AifPushSelectCreationRuleComponent
                    ],
                    bootstrap: [
                        AifPushSelectCreationRuleComponent
                    ],
                    entryComponents: [
                        AifPushSelectCreationRuleComponent
                    ],
                    exports: []
                },] }
    ];
    return AifPushSelectCreationRuleModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { AifAssociationEntity, GetCreationRuleArgs, AifDocumentsTrackManagerModuleConfig, AifPullSelectCreationRuleModuleConfig, AifPushSelectCreationRuleModuleConfig, AifDocumentsTrackManagerComponent, AifDocumentsTrackService, AifDocumentsTrackManagerModule, AifDocumentsPropertyTextMap, TrackDirectionTypeMap, TrackTypeTextMap, TrackType, AifPullSelectCreationRuleComponent, AifPullSelectCreationRuleInternalService, AifPullSelectCreationRuleService, AifPullSelectCreationRuleModule, AifPushSelectCreationRuleComponent, AifPushSelectCreationRuleInternalService, AifPushSelectCreationRuleService, AifPushSelectCreationRuleModule, AifDocumentsTreeComponent as ɵa };
