// @edp-bif/runtime-api/esm2015/lib/entity/tracking/BillDetailInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/BillDetailInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
export class BillDetailInfo extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.detailIds = [];
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.detailCode = null;
        this.detailIds = [];
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "detailCode", this.detailCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "detailIds", BpmnModelJsonHelper.StringListToJson(this.detailIds));
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.detailCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "detailCode");
        this.detailIds = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "detailIds"));
    }
}
if (false) {
    /** @type {?} */
    BillDetailInfo.prototype.detailCode;
    /** @type {?} */
    BillDetailInfo.prototype.detailIds;
}
