// @edp-bif/runtime-api/esm2015/lib/entity/tracking/BillItemInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/BillItemInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { BillDetailInfo } from "./BillDetailInfo";
export class BillItemInfo extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.billDetailInfos = [];
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.itemId = null;
        this.billDetailInfos = [];
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "itemId", this.itemId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billDetailInfos", this.billDetailInfos);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.itemId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "itemId");
        this.billDetailInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "billDetailInfos", BillDetailInfo);
    }
}
if (false) {
    /** @type {?} */
    BillItemInfo.prototype.itemId;
    /** @type {?} */
    BillItemInfo.prototype.billDetailInfos;
}
