// @edp-bif/runtime-api/esm2015/lib/entity/tracking/BillDataInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/BillDataInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { BillChildInfo } from './BillChildInfo';
import { TrackingMappingAssociation } from './TrackingMappingAssociation';
export class BillDataInfo extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.billChildInfos = [];
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.dataId = null;
        this.caption = null;
        this.summary = null;
        this.canContinueTracking = false;
        this.trackingMappingAssociations = null;
        this.billChildInfos = [];
        this.curYear = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "dataId", this.dataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "caption", this.caption);
        BpmnModelJsonHelper.SetAttributeValue(obj, "summary", this.summary);
        BpmnModelJsonHelper.SetAttributeValue(obj, "canContinueTracking", this.canContinueTracking);
        BpmnModelJsonHelper.SetAttributeValue(obj, "trackingMappingAssociations", this.trackingMappingAssociations);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billChildInfos", this.billChildInfos);
        BpmnModelJsonHelper.SetAttributeValue(obj, "curYear", this.curYear);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.dataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dataId");
        this.caption = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "caption");
        this.summary = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "summary");
        this.canContinueTracking = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "canContinueTracking");
        this.trackingMappingAssociations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "trackingMappingAssociations", TrackingMappingAssociation);
        this.billChildInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "billChildInfos", BillChildInfo);
        this.curYear = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "curYear");
    }
}
if (false) {
    /** @type {?} */
    BillDataInfo.prototype.dataId;
    /** @type {?} */
    BillDataInfo.prototype.caption;
    /** @type {?} */
    BillDataInfo.prototype.summary;
    /** @type {?} */
    BillDataInfo.prototype.canContinueTracking;
    /** @type {?} */
    BillDataInfo.prototype.trackingMappingAssociations;
    /** @type {?} */
    BillDataInfo.prototype.billChildInfos;
    /** @type {?} */
    BillDataInfo.prototype.curYear;
}
