// @edp-bif/runtime-api/esm2015/lib/entity/tracking/BillChildInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/BillChildInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { BillItemInfo } from "./BillItemInfo";
export class BillChildInfo extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.billItemInfos = [];
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.itemCode = null;
        this.billItemInfos = [];
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "itemCode", this.itemCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billItemInfos", this.billItemInfos);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.itemCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "itemCode");
        this.billItemInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "billItemInfos", BillItemInfo);
    }
}
if (false) {
    /** @type {?} */
    BillChildInfo.prototype.itemCode;
    /** @type {?} */
    BillChildInfo.prototype.billItemInfos;
}
