// @edp-bif/runtime-api/esm2015/lib/entity/tracking/CanTrackingInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/CanTrackingInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { JointQueryMode } from '@edp-bif/common-api';
import { BillChildInfo } from './BillChildInfo';
/**
 * 此类在graph组件中需要硬转为BillInfo对象，故增减字段时需要同步修改data-service中的build方法
 * @see { BillTrackerDataService }
 */
export class CanTrackingInfo extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.billChildInfos = [];
        this.jointQueryMode = JointQueryMode.DEFAULT;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.canUpTracking = false;
        this.canDownTracking = false;
        this.documentTypeId = null;
        this.bizKind = null;
        this.bizType = null;
        this.dataId = null;
        this.caption = null;
        this.summary = null;
        this.suCode = null;
        this.schemaId = null;
        this.billChildInfos = [];
        this.jointQueryMode = JointQueryMode.DEFAULT;
        this.jointQueryParameters = null;
        this.trackingSourceType = null;
        this.curYear = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "canUpTracking", this.canUpTracking);
        BpmnModelJsonHelper.SetAttributeValue(obj, "canDownTracking", this.canDownTracking);
        BpmnModelJsonHelper.SetAttributeValue(obj, "documentTypeId", this.documentTypeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizKind", this.bizKind);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizType", this.bizType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dataId", this.dataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "caption", this.caption);
        BpmnModelJsonHelper.SetAttributeValue(obj, "summary", this.summary);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaId", this.schemaId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billChildInfos", this.billChildInfos);
        BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryMode", this.jointQueryMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryParameters", this.jointQueryParameters);
        BpmnModelJsonHelper.SetAttributeValue(obj, "trackingSourceType", this.trackingSourceType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "curYear", this.curYear);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.canUpTracking = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "canUpTracking");
        this.canDownTracking = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "canDownTracking");
        this.documentTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "documentTypeId");
        this.bizKind = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizKind");
        this.bizType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizType");
        this.dataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dataId");
        this.caption = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "caption");
        this.summary = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "summary");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.schemaId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaId");
        this.billChildInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "billChildInfos", BillChildInfo);
        this.jointQueryMode = JointQueryMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryMode")];
        this.jointQueryParameters = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryParameters");
        this.trackingSourceType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "trackingSourceType");
        this.curYear = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "curYear");
    }
}
if (false) {
    /** @type {?} */
    CanTrackingInfo.prototype.canUpTracking;
    /** @type {?} */
    CanTrackingInfo.prototype.canDownTracking;
    /** @type {?} */
    CanTrackingInfo.prototype.documentTypeId;
    /** @type {?} */
    CanTrackingInfo.prototype.bizKind;
    /** @type {?} */
    CanTrackingInfo.prototype.bizType;
    /** @type {?} */
    CanTrackingInfo.prototype.dataId;
    /** @type {?} */
    CanTrackingInfo.prototype.caption;
    /** @type {?} */
    CanTrackingInfo.prototype.summary;
    /** @type {?} */
    CanTrackingInfo.prototype.suCode;
    /** @type {?} */
    CanTrackingInfo.prototype.schemaId;
    /** @type {?} */
    CanTrackingInfo.prototype.billChildInfos;
    /** @type {?} */
    CanTrackingInfo.prototype.jointQueryMode;
    /** @type {?} */
    CanTrackingInfo.prototype.jointQueryParameters;
    /** @type {?} */
    CanTrackingInfo.prototype.trackingSourceType;
    /** @type {?} */
    CanTrackingInfo.prototype.curYear;
}
