// @edp-bif/runtime-api/esm2015/lib/entity/tracking/DecimalTypeFormatInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/DecimalTypeFormatInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { ScaleMode } from "./ScaleMode";
export class DecimalTypeFormatInfo extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.scaleMode = ScaleMode.DEFAULT;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.scale = 0;
        this.scaleMode = ScaleMode.DEFAULT;
        this.fieldPath = null;
        this.version = 0;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "scale", this.scale);
        BpmnModelJsonHelper.SetAttributeValue(obj, "scale", this.scale);
        BpmnModelJsonHelper.SetAttributeValue(obj, "scaleMode", this.scaleMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "fieldPath", this.fieldPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "version", this.version);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.scale = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "scale");
        /** @type {?} */
        const scaleMode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "scaleMode");
        this.scaleMode = scaleMode == null ? ScaleMode.DEFAULT : ScaleMode[scaleMode];
        this.fieldPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "fieldPath");
        this.version = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "version");
    }
}
if (false) {
    /** @type {?} */
    DecimalTypeFormatInfo.prototype.scale;
    /** @type {?} */
    DecimalTypeFormatInfo.prototype.scaleMode;
    /** @type {?} */
    DecimalTypeFormatInfo.prototype.fieldPath;
    /** @type {?} */
    DecimalTypeFormatInfo.prototype.version;
}
