// @edp-bif/runtime-api/esm2015/lib/entity/tracking/GetBillsInfoArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/GetBillsInfoArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { BillChildInfo } from './BillChildInfo';
export class GetBillsInfoArgs extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.billChildInfos = [];
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.documentTypeId = null;
        this.bizKind = null;
        this.bizType = null;
        this.billDataId = null;
        this.billChildInfos = [];
        this.curYear = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "documentTypeId", this.documentTypeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizKind", this.bizKind);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizType", this.bizType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billDataId", this.billDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billChildInfos", this.billChildInfos);
        BpmnModelJsonHelper.SetAttributeValue(obj, "curYear", this.curYear);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.documentTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "documentTypeId");
        this.bizKind = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizKind");
        this.bizType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizType");
        this.billDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billDataId");
        this.billChildInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "billChildInfos", BillChildInfo);
        this.curYear = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "curYear");
    }
}
if (false) {
    /** @type {?} */
    GetBillsInfoArgs.prototype.documentTypeId;
    /** @type {?} */
    GetBillsInfoArgs.prototype.bizKind;
    /** @type {?} */
    GetBillsInfoArgs.prototype.bizType;
    /** @type {?} */
    GetBillsInfoArgs.prototype.billDataId;
    /** @type {?} */
    GetBillsInfoArgs.prototype.billChildInfos;
    /** @type {?} */
    GetBillsInfoArgs.prototype.curYear;
}
