// @edp-bif/runtime-api/esm2015/lib/entity/tracking/JointQueryParametersContent.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/JointQueryParametersContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { JointQueryParameterMode } from "./JointQueryParameterMode";
export class JointQueryParametersContent extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.jointQueryParameterMode = JointQueryParameterMode.FIELD;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.jointQueryParameterMode = JointQueryParameterMode.FIELD;
        this.content = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryParameterMode", this.jointQueryParameterMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.jointQueryParameterMode = JointQueryParameterMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryParameterMode")];
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
    }
}
if (false) {
    /** @type {?} */
    JointQueryParametersContent.prototype.jointQueryParameterMode;
    /** @type {?} */
    JointQueryParametersContent.prototype.content;
}
