// @edp-bif/runtime-api/esm2015/lib/entity/tracking/SimpleValueMappingInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/SimpleValueMappingInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
export class SimpleValueMappingInfo extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.displayValue = null;
        this.actualValue = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "displayValue", this.displayValue);
        BpmnModelJsonHelper.SetAttributeValue(obj, "actualValue", this.actualValue);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.displayValue = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "displayValue");
        this.actualValue = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "actualValue");
    }
}
if (false) {
    /** @type {?} */
    SimpleValueMappingInfo.prototype.displayValue;
    /** @type {?} */
    SimpleValueMappingInfo.prototype.actualValue;
}
