// @edp-bif/runtime-api/esm2015/lib/entity/tracking/StringTypeFormatInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/StringTypeFormatInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { SimpleValueMappingInfo } from "./SimpleValueMappingInfo";
import { StringDisplayMode } from "./StringDisplayMode";
export class StringTypeFormatInfo extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.stringDisplayMode = StringDisplayMode.SIMPLEVALUEMAPPING;
        this.simpleValueMappingInfos = [];
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.stringDisplayMode = StringDisplayMode.SIMPLEVALUEMAPPING;
        this.simpleValueMappingInfos = [];
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "stringDisplayMode", this.stringDisplayMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "simpleValueMappingInfos", this.simpleValueMappingInfos);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.stringDisplayMode = StringDisplayMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "stringDisplayMode")];
        this.simpleValueMappingInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "simpleValueMappingInfos", SimpleValueMappingInfo);
    }
}
if (false) {
    /** @type {?} */
    StringTypeFormatInfo.prototype.stringDisplayMode;
    /** @type {?} */
    StringTypeFormatInfo.prototype.simpleValueMappingInfos;
}
