// @edp-bif/runtime-api/esm2015/lib/entity/tracking/SummaryFieldsInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/SummaryFieldsInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { FieldsWithOrder } from "./FieldsWithOrder";
export class SummaryFieldsInfo extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.fields = [];
        this.childFieldInfos = [];
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.code = null;
        this.fields = [];
        this.childFieldInfos = [];
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "fields", this.fields);
        BpmnModelJsonHelper.SetAttributeValue(obj, "childFieldInfos", this.childFieldInfos);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.fields = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "fields", FieldsWithOrder);
        this.childFieldInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "childFieldInfos", SummaryFieldsInfo);
    }
}
if (false) {
    /** @type {?} */
    SummaryFieldsInfo.prototype.code;
    /** @type {?} */
    SummaryFieldsInfo.prototype.fields;
    /** @type {?} */
    SummaryFieldsInfo.prototype.childFieldInfos;
}
