// @edp-bif/runtime-api/esm2015/lib/entity/tracking/TrackingMappingAssociation.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/TrackingMappingAssociation.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
export class TrackingMappingAssociation extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.upBillTableCodePath = null;
        this.upBillDataIdPath = null;
        this.downBillTableCodePath = null;
        this.downBillDataIdPath = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTableCodePath", this.upBillTableCodePath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillDataIdPath", this.upBillDataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTableCodePath", this.downBillTableCodePath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillDataIdPath", this.downBillDataIdPath);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.upBillTableCodePath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTableCodePath");
        this.upBillDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillDataIdPath");
        this.downBillTableCodePath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTableCodePath");
        this.downBillDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillDataIdPath");
    }
}
if (false) {
    /** @type {?} */
    TrackingMappingAssociation.prototype.upBillTableCodePath;
    /** @type {?} */
    TrackingMappingAssociation.prototype.upBillDataIdPath;
    /** @type {?} */
    TrackingMappingAssociation.prototype.downBillTableCodePath;
    /** @type {?} */
    TrackingMappingAssociation.prototype.downBillDataIdPath;
}
