// @farris/ui-datagrid-settings/esm5/lib/multi-headers/preview/show-preview.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, HostListener, ComponentFactoryResolver } from "@angular/core";
import { ColumnTreeComponent } from "../column-tree.component";
import { BsModalService } from "@farris/ui-modal";
import { PreviewMultiHeadersComponent } from "./preview-multi-headers.component";
var ShowPreviewDirective = /** @class */ (function () {
    function ShowPreviewDirective(el, cfr, modalSer) {
        this.el = el;
        this.cfr = cfr;
        this.modalSer = modalSer;
        this.dlgTitle = '效果预览';
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    ShowPreviewDirective.prototype.onPreviewBtnClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var modalRef = this.modalSer.show(PreviewMultiHeadersComponent, {
            title: this.dlgTitle,
            width: 600,
            height: 300,
            showMaxButton: true,
            resizable: true,
            initialState: {
                previewGroupColumns: this.treeIns.getPreviewGroupColumns()
            },
            showButtons: false
        });
    };
    ShowPreviewDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[preview-headers]'
                },] }
    ];
    /** @nocollapse */
    ShowPreviewDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ComponentFactoryResolver },
        { type: BsModalService }
    ]; };
    ShowPreviewDirective.propDecorators = {
        treeIns: [{ type: Input, args: ['preview-headers',] }],
        dlgTitle: [{ type: Input }],
        onPreviewBtnClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return ShowPreviewDirective;
}());
export { ShowPreviewDirective };
if (false) {
    /** @type {?} */
    ShowPreviewDirective.prototype.treeIns;
    /** @type {?} */
    ShowPreviewDirective.prototype.dlgTitle;
    /**
     * @type {?}
     * @private
     */
    ShowPreviewDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    ShowPreviewDirective.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    ShowPreviewDirective.prototype.modalSer;
}
