// @farris/ui-lookup/esm2015/lib/services/lookup-selection.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { of, BehaviorSubject, Subject } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { FavoriteAction } from '../lookup-displaytype';
/**
 * @record
 */
export function SelectionState() { }
if (false) {
    /** @type {?} */
    SelectionState.prototype.selecteditems;
    /** @type {?} */
    SelectionState.prototype.favoriteItems;
    /** @type {?} */
    SelectionState.prototype.quickItems;
}
export class LookupSelectionService {
    /**
     * @param {?} ins
     */
    constructor(ins) {
        this.ins = ins;
        this.state = {
            selecteditems: [],
            favoriteItems: [],
            quickItems: []
        };
        this.state$ = new BehaviorSubject(this.state);
        this.selected$ = this.state$.pipe(switchMap((/**
         * @param {?} n
         * @return {?}
         */
        n => of(n.selecteditems))));
        this.favoriteItems$ = new BehaviorSubject({ items: this.state.favoriteItems, action: null });
        this.quickItems$ = this.state$.pipe(switchMap((/**
         * @param {?} n
         * @return {?}
         */
        n => of(n.quickItems))));
        this.selectionChanged$ = new Subject();
    }
    /**
     * @private
     * @return {?}
     */
    get idField() {
        return this.ins.idField;
    }
    /**
     * @param {?} items
     * @return {?}
     */
    initFavoriteItems(items) {
        this.state.favoriteItems = items || [];
    }
    //#region 收藏数据
    /**
     * @param {?} data
     * @param {?} action
     * @return {?}
     */
    updateFavoriteData(data, action) {
        if (this.ins.savingFaoriteData) {
            return;
        }
        if (action === FavoriteAction.add) {
            this.state.favoriteItems = this.state.favoriteItems.concat([data]);
        }
        else {
            this.state.favoriteItems = this.state.favoriteItems.filter((/**
             * @param {?} n
             * @return {?}
             */
            n => n[this.idField] !== data[this.idField]));
        }
        this.favoriteItems$.next({ items: this.state.favoriteItems, action, data });
    }
    //#endregion
    //#region 多选数据
    /**
     * @param {?} data
     * @return {?}
     */
    loadSelections(data) {
        this.state.selecteditems = [...data];
        this.state$.next(this.state);
    }
    /**
     * @return {?}
     */
    getSelections() {
        return [...this.state.selecteditems];
    }
    /**
     * @param {?} item
     * @return {?}
     */
    select(item) {
        if (item) {
            this.state.selecteditems = [...this.state.selecteditems, item];
            this.state$.next(this.state);
            this.selectionChanged$.next({ data: [item[this.idField]], selected: true });
        }
    }
    /**
     * @param {?} pathcode
     * @return {?}
     */
    unselectByPathcode(pathcode) {
        /** @type {?} */
        const ids = [];
        this.state.selecteditems = this.state.selecteditems.filter((/**
         * @param {?} n
         * @return {?}
         */
        n => {
            /** @type {?} */
            const _pathcode = this.ins.getPathCode(n, this.ins.treeInfo);
            /** @type {?} */
            const r = _pathcode && _pathcode.indexOf(pathcode) !== 0;
            if (!r) {
                ids.push(n[this.idField]);
            }
            return r;
        }));
        this.selectionChanged$.next({ data: ids, selected: false });
    }
    /**
     * @param {?} data
     * @param {?=} checked
     * @return {?}
     */
    updateSelections(data, checked = true) {
        if (!Array.isArray(data)) {
            data = [data];
        }
        /** @type {?} */
        const items = [...data];
        /** @type {?} */
        const idField = this.idField;
        /** @type {?} */
        const ids = items.map((/**
         * @param {?} n
         * @return {?}
         */
        n => n[idField]));
        if (checked) {
            if (this.state.selecteditems && !this.state.selecteditems.length) {
                this.state.selecteditems = items;
            }
            else {
                ids.forEach((/**
                 * @param {?} n
                 * @param {?} i
                 * @return {?}
                 */
                (n, i) => {
                    if (!this.state.selecteditems.find((/**
                     * @param {?} r
                     * @return {?}
                     */
                    r => r[idField] == n))) {
                        this.state.selecteditems.push(items[i]);
                    }
                }));
            }
        }
        else {
            if (data) {
                this.state.selecteditems = this.state.selecteditems.filter((/**
                 * @param {?} n
                 * @return {?}
                 */
                n => {
                    return ids.indexOf(n[idField]) === -1;
                }));
            }
        }
        this.state$.next(this.state);
        this.selectionChanged$.next({ data: ids, selected: checked });
    }
    /**
     * @param {?} id
     * @return {?}
     */
    unSelect(id) {
        if (id) {
            if (Array.isArray(id)) {
                this.state.selecteditems = this.state.selecteditems.filter((/**
                 * @param {?} n
                 * @return {?}
                 */
                n => {
                    return id.indexOf(n[this.idField]) === -1;
                }));
                this.selectionChanged$.next({ data: id, selected: false });
            }
            else {
                this.state.selecteditems = this.state.selecteditems.filter((/**
                 * @param {?} n
                 * @return {?}
                 */
                n => n[this.idField] != id));
                this.selectionChanged$.next({ data: [id], selected: false });
            }
            this.state$.next(this.state);
        }
    }
    /**
     * @return {?}
     */
    clearSelections() {
        this.state.selecteditems = [];
        this.state$.next(this.state);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    LookupSelectionService.prototype.state;
    /** @type {?} */
    LookupSelectionService.prototype.state$;
    /** @type {?} */
    LookupSelectionService.prototype.selected$;
    /** @type {?} */
    LookupSelectionService.prototype.favoriteItems$;
    /** @type {?} */
    LookupSelectionService.prototype.quickItems$;
    /** @type {?} */
    LookupSelectionService.prototype.selectionChanged$;
    /**
     * @type {?}
     * @private
     */
    LookupSelectionService.prototype.ins;
}
