// @edp-bif/runtime-api/esm2015/lib/bif-manual-creation/args/PushTaskArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bif-manual-creation/args/PushTaskArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { RpcCreationRuleExecutorResult } from "@edp-aif/runtime-api";
export class PushTaskArgs extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.suCode = null;
        this.creationRuleExecutorResult = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleExecutorResult", this.creationRuleExecutorResult);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.creationRuleExecutorResult = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "creationRuleExecutorResult", RpcCreationRuleExecutorResult);
    }
}
if (false) {
    /** @type {?} */
    PushTaskArgs.prototype.suCode;
    /** @type {?} */
    PushTaskArgs.prototype.creationRuleExecutorResult;
}
