// @farris/mobile-devkit/esm5/lib/expression/listener/ui_state_change_listener.js
import * as tslib_1 from "tslib";
import { ChangeListener } from './change_listener';
import { Expression } from '../expression/index';
import { NAMESPACE } from '../core/index';
/**
 * 监听UIState变更
 */
var UIStateChangeListener = /** @class */ (function (_super) {
    tslib_1.__extends(UIStateChangeListener, _super);
    /**
     * 构造函数
     */
    function UIStateChangeListener(viewModelContext) {
        var _this = _super.call(this) || this;
        _this.viewModelContext = viewModelContext;
        _this.frameId = _this.viewModelContext.id;
        _this.uiState = _this.viewModelContext.uiState;
        _this.namespace = _this.viewModelContext.injector.get(NAMESPACE, null);
        _this.registerEvent();
        return _this;
    }
    UIStateChangeListener.prototype.buildEventPath = function (change) {
        return null;
    };
    UIStateChangeListener.prototype.registerEvent = function () {
        var _this = this;
        if (this.uiState && this.uiState.changes) {
            this.uiState.changes.subscribe(function (change) {
                var modification = {
                    ns: _this.namespace,
                    path: [change.field],
                    type: Expression.EventType.ValueChanged,
                    value: change.value,
                    source: Expression.EventSource.State,
                    frameId: _this.frameId
                };
                // console.log("UIStateChangeListener", modification);
                _this.subject.next(modification);
            });
        }
    };
    return UIStateChangeListener;
}(ChangeListener));
export { UIStateChangeListener };
