// @gsp-dip/data-imp-exp/esm2015/lib/model/Model/DataExchangeObject.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DataExchangeElement } from './DataExchangeElement';
import { DemSource } from './DemSource';
import { RepeatBasis } from '../InputStep/RepeatBasis/repeat-basis';
/**
 * 数据转换对象
 */
export class DataExchangeObject {
    constructor() {
        /**
         * 父对象ID
         */
        this.ParentObjectId = '';
        /**
         * 文件中对象名称
         */
        this.FileObjectName = '';
        /**
         * 是否导出
         */
        this.Export = true;
        /**
         * 序号
         */
        this.Order = 1;
        /**
         * 对象字段集合
         */
        this.Elements = new Array();
        /**
         * 子对象集合
         */
        this.ChildrenDeo = new Array();
        /**
         * 数据来源
         */
        this.Source = new DemSource();
        /**
         * 重复依据
         */
        this.RepeatBases = new Array();
    }
    /**
     * copy
     * @param {?} obj
     * @return {?}
     */
    static copy(obj) {
        /** @type {?} */
        const deo = new DataExchangeObject();
        deo.Export = obj.Export;
        deo.ID = obj.ID;
        deo.Code = obj.Code;
        deo.Name = obj.Name;
        deo.Source = obj.Source;
        deo.ParentObjectId = obj.ParentObjectId;
        deo.FileObjectName = obj.FileObjectName;
        deo.Order = obj.Order;
        deo.RefObjectName = obj.RefObjectName;
        deo.Elements = new Array();
        for (const ele of obj.Elements) {
            deo.Elements.push(DataExchangeElement.copy(ele));
        }
        deo.ChildrenDeo = new Array();
        for (const item of obj.ChildrenDeo) {
            deo.ChildrenDeo.push(DataExchangeObject.copy(item));
        }
        deo.RepeatBases = new Array();
        for (const item of obj.RepeatBases) {
            deo.RepeatBases.push(RepeatBasis.copy(item));
        }
        return deo;
    }
}
if (false) {
    /**
     * 父对象ID
     * @type {?}
     */
    DataExchangeObject.prototype.ParentObjectId;
    /**
     * 文件中对象名称
     * @type {?}
     */
    DataExchangeObject.prototype.FileObjectName;
    /**
     * 是否导出
     * @type {?}
     */
    DataExchangeObject.prototype.Export;
    /**
     * 序号
     * @type {?}
     */
    DataExchangeObject.prototype.Order;
    /**
     * 对象字段集合
     * @type {?}
     */
    DataExchangeObject.prototype.Elements;
    /**
     * 子对象集合
     * @type {?}
     */
    DataExchangeObject.prototype.ChildrenDeo;
    /**
     * 数据来源
     * @type {?}
     */
    DataExchangeObject.prototype.Source;
    /**
     * 重复依据
     * @type {?}
     */
    DataExchangeObject.prototype.RepeatBases;
    /** @type {?} */
    DataExchangeObject.prototype.ID;
    /** @type {?} */
    DataExchangeObject.prototype.Code;
    /** @type {?} */
    DataExchangeObject.prototype.Name;
    /** @type {?} */
    DataExchangeObject.prototype.RefObjectName;
}
