// @edp-pmf/bpmn-model/dist/esm/lib/model/omgdc/Bounds.js
import { __extends } from "tslib";
import { BpmnModelHelper } from "../BpmnModelHelper";
import { OmgdcElement } from "./OmgdcElement";
var Bounds = /** @class */ (function (_super) {
    __extends(Bounds, _super);
    function Bounds(model) {
        return _super.call(this, model) || this;
    }
    Bounds.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Height = 0;
        this.Width = 0;
        this.X = 0;
        this.Y = 0;
    };
    Bounds.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "Height", this.Height);
        BpmnModelHelper.SetAttributeValue(obj, "Width", this.Width);
        BpmnModelHelper.SetAttributeValue(obj, "X", this.X);
        BpmnModelHelper.SetAttributeValue(obj, "Y", this.Y);
        return obj;
    };
    Bounds.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Height = obj.Height;
        this.Width = obj.Width;
        this.X = obj.X;
        this.Y = obj.Y;
    };
    return Bounds;
}(OmgdcElement));
export { Bounds };
