// @farris/ui-datagrid-settings/esm5/lib/multi-headers/new-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, ElementRef } from '@angular/core';
var NewHeaderFormComponent = /** @class */ (function () {
    function NewHeaderFormComponent() {
        this.label = '表头名称';
        this.placeholder = "输入表头名称";
    }
    /**
     * @return {?}
     */
    NewHeaderFormComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    NewHeaderFormComponent.decorators = [
        { type: Component, args: [{
                    selector: 'group-header-form',
                    template: "\n    <div class=\"d-flex flex-row\" style=\"align-items: center;height: 100%;\">\n        <span style=\"width: 90px;text-align: right;\">{{label}}</span>\n        <input #txt placeholder=\"{{placeholder}}\" type=\"text\" [(ngModel)]=\"title\" class=\"form-control ml-2\" style=\"width: 210px\">\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    NewHeaderFormComponent.ctorParameters = function () { return []; };
    NewHeaderFormComponent.propDecorators = {
        label: [{ type: Input }],
        title: [{ type: Input }],
        placeholder: [{ type: Input }],
        input: [{ type: ViewChild, args: ['txt',] }]
    };
    return NewHeaderFormComponent;
}());
export { NewHeaderFormComponent };
if (false) {
    /** @type {?} */
    NewHeaderFormComponent.prototype.label;
    /** @type {?} */
    NewHeaderFormComponent.prototype.title;
    /** @type {?} */
    NewHeaderFormComponent.prototype.placeholder;
    /** @type {?} */
    NewHeaderFormComponent.prototype.input;
}
