// @farris/ui-datagrid-settings/esm5/lib/multi-headers/setcolumns.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from "@angular/core";
var SetColumnsComponent = /** @class */ (function () {
    function SetColumnsComponent() {
        this.data = [];
        this.checkedItems = [];
        this.selectedIds = [];
    }
    /**
     * @return {?}
     */
    SetColumnsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.selectedIds = (this.checkedItems || []).map((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n.id; }));
        if (this.data) {
            this.data.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (!item.id) {
                    item.id = item.field;
                }
            }));
        }
    };
    SetColumnsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'set-columns',
                    template: "<multi-select [dataSource]=\"data\"\r\n[idField]=\"'id'\" [valueField]=\"'field'\" [textField]=\"'title'\" [displayType]=\"'List'\"\r\n[enableTargetSort]=\"true\" [showCheckbox]=\"true\" [enableMoveMore]=\"true\" [(selectedId)]=\"selectedIds\">\r\n</multi-select>"
                }] }
    ];
    /** @nocollapse */
    SetColumnsComponent.ctorParameters = function () { return []; };
    SetColumnsComponent.propDecorators = {
        data: [{ type: Input }],
        checkedItems: [{ type: Input }]
    };
    return SetColumnsComponent;
}());
export { SetColumnsComponent };
if (false) {
    /** @type {?} */
    SetColumnsComponent.prototype.data;
    /** @type {?} */
    SetColumnsComponent.prototype.checkedItems;
    /** @type {?} */
    SetColumnsComponent.prototype.selectedIds;
}
