// @farris/ui-forms/esm2015/lib/shortcut-tab.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventManager } from '@angular/platform-browser';
import { Directive, ElementRef, Injector } from '@angular/core';
export class ShortCutTabKeyDirective {
    /**
     * @param {?} el
     * @param {?} injector
     * @param {?} eventManager
     */
    constructor(el, injector, eventManager) {
        this.el = el;
        this.injector = injector;
        this.eventManager = eventManager;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.eventManager.addEventListener(this.el.nativeElement, 'keydown', (/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            e.stopPropagation();
            if (e.key === 'Tab') {
                this.onTabKeydownHandler(e.target);
            }
        }));
    }
    /**
     * @private
     * @param {?} currentTarget
     * @return {?}
     */
    onTabKeydownHandler(currentTarget) {
        /** @type {?} */
        let allInputs = this.el.nativeElement.querySelectorAll('input,textarea');
        allInputs = Array.from(allInputs);
        /** @type {?} */
        const currentInputIndex = allInputs.findIndex((/**
         * @param {?} n
         * @return {?}
         */
        n => n === currentTarget));
        /** @type {?} */
        let nextInputIndex = currentInputIndex + 1;
        if (nextInputIndex >= allInputs.length) {
            nextInputIndex = 0;
        }
        /** @type {?} */
        const nextInputTarget = allInputs[nextInputIndex];
        if (nextInputTarget) {
            setTimeout((/**
             * @return {?}
             */
            () => {
                nextInputTarget.focus();
            }));
        }
    }
}
ShortCutTabKeyDirective.decorators = [
    { type: Directive, args: [{
                selector: 'shortcut-tab',
            },] }
];
/** @nocollapse */
ShortCutTabKeyDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Injector },
    { type: EventManager }
];
if (false) {
    /** @type {?} */
    ShortCutTabKeyDirective.prototype.el;
    /** @type {?} */
    ShortCutTabKeyDirective.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    ShortCutTabKeyDirective.prototype.eventManager;
}
