// @farris/devkit/esm2015/lib/frame/component_ref_manager.js
/**
 * 组件实例管理
 */
export class ComponentRefManager {
    constructor() {
        this.components = new Map();
    }
    /**
     * 增加组件实例
     * @param id
     * @param component
     */
    add(id, component) {
        const components = this.components.get(id);
        if (components) {
            components.push(component);
        }
        else {
            this.components.set(id, [component]);
        }
    }
    /**
     * 删除组件实例
     * @param id
     */
    remove(id) {
        const components = this.components.get(id);
        if (components && components.length > 0) {
            components.length = 0;
            this.components.delete(id);
        }
    }
    /**
     * 通过id获取组件实例
     * @param id
     * @returns
     */
    get(id) {
        if (id) {
            const components = this.components.get(id);
            if (components && components.length > 0) {
                return components;
            }
        }
        return null;
    }
    /**
     * 是否包含id的组件
     * @param id
     * @returns
     */
    has(id) {
        return this.components.has(id);
    }
    /**
     * 清空
     */
    clear() {
        this.components.clear();
    }
    /**
     * 获取指定类型的组件
     * @param componentType
     * @returns
     */
    getComponentByType(componentType) {
        const containers = Array.from(this.components.values());
        let component = null;
        for (let containerIndex = 0; containers && containerIndex < containers.length; containerIndex++) {
            const components = containers[containerIndex];
            for (let index = 0; components && index < components.length; index++) {
                if (components[index] instanceof componentType) {
                    component = components[index];
                    break;
                }
            }
        }
        return component;
    }
}
