// @farris/ui-enum-editor/esm5/lib/enum-editor-footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, HostBinding, Output, Input } from '@angular/core';
import { EnumEditorComponent } from './enum-editor.component';
var EnumEditorFooterComponent = /** @class */ (function () {
    function EnumEditorFooterComponent(grid) {
        this.grid = grid;
        this.cls = 'f-filter-footer';
        this.showSortBtns = true;
        this.addFilter = new EventEmitter();
        this.clearFilter = new EventEmitter();
        this.moveTop = new EventEmitter();
        this.movePrev = new EventEmitter();
        this.moveNext = new EventEmitter();
        this.moveBottom = new EventEmitter();
    }
    /**
     * @return {?}
     */
    EnumEditorFooterComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    EnumEditorFooterComponent.prototype.onAddFilter = /**
     * @return {?}
     */
    function () {
        this.addFilter.emit();
    };
    /**
     * @return {?}
     */
    EnumEditorFooterComponent.prototype.onClear = /**
     * @return {?}
     */
    function () {
        this.clearFilter.emit();
    };
    /**
     * @return {?}
     */
    EnumEditorFooterComponent.prototype.onMoveTop = /**
     * @return {?}
     */
    function () {
        this.moveTop.emit();
    };
    /**
     * @return {?}
     */
    EnumEditorFooterComponent.prototype.onMovePrev = /**
     * @return {?}
     */
    function () {
        this.movePrev.emit();
    };
    /**
     * @return {?}
     */
    EnumEditorFooterComponent.prototype.onMoveNext = /**
     * @return {?}
     */
    function () {
        this.moveNext.emit();
    };
    /**
     * @return {?}
     */
    EnumEditorFooterComponent.prototype.onMoveBottom = /**
     * @return {?}
     */
    function () {
        this.moveBottom.emit();
    };
    /**
     * @return {?}
     */
    EnumEditorFooterComponent.prototype.canUse = /**
     * @return {?}
     */
    function () {
        if (this.grid.currentEnumData) {
            return this.grid.data.length > 1;
        }
        return false;
    };
    /**
     * @return {?}
     */
    EnumEditorFooterComponent.prototype.canMoveUp = /**
     * @return {?}
     */
    function () {
        if (this.canUse()) {
            return this.grid.currentEnumData.index > 0;
        }
        return false;
    };
    /**
     * @return {?}
     */
    EnumEditorFooterComponent.prototype.canMoveDown = /**
     * @return {?}
     */
    function () {
        if (this.canUse()) {
            return this.grid.currentEnumData.index < this.grid.data.length - 1;
        }
        return false;
    };
    EnumEditorFooterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'enum-editor-footer',
                    template: "\n    <div class=\"footer-container px-2\" style=\"height: 36px;\">\n        <button type=\"button\" (click)=\"onAddFilter()\"\n            class=\"k-button k-button-icontext k-flat \"><span class=\"k-icon k-i-add\"></span> {{ 'enumEditor.addWhere' | locale }}</button>\n        <button type=\"button\" (click)=\"onClear()\" [disabled]=\"!grid.data.length\"\n            class=\"k-button k-button-icontext k-flat \"><span class=\"k-icon k-i-delete\"></span> {{ 'enumEditor.clear' | locale }}</button>\n        <ng-container *ngIf=\"showSortBtns\">\n            <button type=\"button\" [disabled]=\"!canMoveUp()\" (click)=\"onMoveTop()\"\n            class=\"k-button k-button-icontext k-flat \"><span class=\"k-icon k-i-arrow-end-up\"></span>\n            {{'enumEditor.moveTop' | locale }}</button>\n        <button type=\"button\" [disabled]=\"!canMoveUp()\" (click)=\"onMovePrev()\"\n            class=\"k-button k-button-icontext k-flat \"><span class=\"k-icon k-i-arrow-60-up\"></span>\n            {{'enumEditor.moveUp' | locale }}</button>\n        <button type=\"button\" [disabled]=\"!canMoveDown()\" (click)=\"onMoveNext()\"\n            class=\"k-button k-button-icontext k-flat \"><span class=\"k-icon k-i-arrow-60-down\"></span>\n            {{'enumEditor.moveDown' | locale }}</button>\n        <button type=\"button\" [disabled]=\"!canMoveDown()\" (click)=\"onMoveBottom()\"\n            class=\"k-button k-button-icontext k-flat\"><span class=\"k-icon k-i-arrow-end-down\"></span>\n            {{'enumEditor.moveBottom' | locale }}</button>\n            </ng-container>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    EnumEditorFooterComponent.ctorParameters = function () { return [
        { type: EnumEditorComponent }
    ]; };
    EnumEditorFooterComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        showSortBtns: [{ type: Input }],
        addFilter: [{ type: Output }],
        clearFilter: [{ type: Output }],
        moveTop: [{ type: Output }],
        movePrev: [{ type: Output }],
        moveNext: [{ type: Output }],
        moveBottom: [{ type: Output }]
    };
    return EnumEditorFooterComponent;
}());
export { EnumEditorFooterComponent };
if (false) {
    /** @type {?} */
    EnumEditorFooterComponent.prototype.cls;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.showSortBtns;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.addFilter;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.clearFilter;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.moveTop;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.movePrev;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.moveNext;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.moveBottom;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.grid;
}
