// @farris/ui-enum-editor/esm5/lib/enum-editor-input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { InputGroupComponent } from '@farris/ui-input-group';
import { EnumEditorService, EnumOutType } from './enum-editor.service';
import { Component, Injector, ElementRef, Input, EventEmitter, Output, forwardRef, ViewChild, ChangeDetectorRef } from '@angular/core';
import { LookupComponent } from '@farris/ui-lookup';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/** @type {?} */
export var ENUM_EDITOR_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef((/**
     * @return {?}
     */
    function () { return EnumEditorInputComponent; })),
    multi: true
};
var EnumEditorInputComponent = /** @class */ (function (_super) {
    tslib_1.__extends(EnumEditorInputComponent, _super);
    function EnumEditorInputComponent(injector, ees, el) {
        var _this = _super.call(this, injector, el) || this;
        _this.injector = injector;
        _this.ees = ees;
        _this.el = el;
        _this.textField = 'name';
        _this.valueField = 'value';
        _this.outType = EnumOutType.array;
        _this.enableClear = true;
        _this.valueChange = new EventEmitter();
        _this.groupIcon = '<i class="f-icon f-icon-lookup"></i>';
        _this.cd = null;
        _this.ees.dataChanged.subscribe((/**
         * @param {?} str
         * @return {?}
         */
        function (str) {
            _this.updateData(str);
        }));
        _this.cd = _this.injector.get(ChangeDetectorRef, null);
        return _this;
    }
    /**
     * @private
     * @param {?} str
     * @return {?}
     */
    EnumEditorInputComponent.prototype.updateData = /**
     * @private
     * @param {?} str
     * @return {?}
     */
    function (str) {
        this.value = str;
        if (this.cd) {
            this.cd.detectChanges();
        }
        this.valueChange.emit(str);
        this.onModelChange(str);
        this.onModelTouched(str);
    };
    /**
     * @return {?}
     */
    EnumEditorInputComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    EnumEditorInputComponent.prototype.showDialog = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var enumData = this.convertValue2JSON();
        this.ees.showDialog(enumData, {
            textField: this.textField,
            valueField: this.valueField,
            outType: this.outType
        });
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    EnumEditorInputComponent.prototype.onClear = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.updateData('');
    };
    /**
     * @param {?} val
     * @return {?}
     */
    EnumEditorInputComponent.prototype.writeValue = /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        if (val && typeof val === 'object') {
            this.value = JSON.stringify(val);
        }
        else {
            this.value = val || '';
        }
    };
    /**
     * @private
     * @return {?}
     */
    EnumEditorInputComponent.prototype.convertValue2JSON = /**
     * @private
     * @return {?}
     */
    function () {
        return this.ees.toJSON(this.value);
    };
    EnumEditorInputComponent.decorators = [
        { type: Component, args: [{
                    selector: 'enum-editor-input',
                    template: "\n    <input-group #ig style=\"width: 100%\"\n        [groupText]=\"groupIcon\"\n        (clear)=\"onClear($event)\"\n        (clickHandle)=\"showDialog()\"\n        [readonly] = \"readonly\"\n        [editable] = \"editable\"\n        [enableClear] = \"enableClear\"\n        [(value)]=\"value\">\n    </input-group>\n    ",
                    providers: [
                        ENUM_EDITOR_VALUE_ACCESSOR,
                        EnumEditorService
                    ]
                }] }
    ];
    /** @nocollapse */
    EnumEditorInputComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: EnumEditorService },
        { type: ElementRef }
    ]; };
    EnumEditorInputComponent.propDecorators = {
        value: [{ type: Input }],
        textField: [{ type: Input }],
        valueField: [{ type: Input }],
        outType: [{ type: Input }],
        enableClear: [{ type: Input }],
        valueChange: [{ type: Output }],
        inputGroupRef: [{ type: ViewChild, args: ['ig',] }]
    };
    return EnumEditorInputComponent;
}(LookupComponent));
export { EnumEditorInputComponent };
if (false) {
    /** @type {?} */
    EnumEditorInputComponent.prototype.value;
    /** @type {?} */
    EnumEditorInputComponent.prototype.textField;
    /** @type {?} */
    EnumEditorInputComponent.prototype.valueField;
    /** @type {?} */
    EnumEditorInputComponent.prototype.outType;
    /** @type {?} */
    EnumEditorInputComponent.prototype.enableClear;
    /** @type {?} */
    EnumEditorInputComponent.prototype.valueChange;
    /** @type {?} */
    EnumEditorInputComponent.prototype.inputGroupRef;
    /** @type {?} */
    EnumEditorInputComponent.prototype.groupIcon;
    /** @type {?} */
    EnumEditorInputComponent.prototype.cd;
    /** @type {?} */
    EnumEditorInputComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    EnumEditorInputComponent.prototype.ees;
    /** @type {?} */
    EnumEditorInputComponent.prototype.el;
}
