// @farris/ui-enum-editor/esm5/lib/enum-editor.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InputGroupModule } from '@farris/ui-input-group';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { MessagerModule } from '@farris/ui-messager';
import { LocaleModule } from '@farris/ui-locale';
import { NgModule } from '@angular/core';
import { EnumEditorInputComponent } from './enum-editor-input.component';
import { EnumEditorComponent } from './enum-editor.component';
import { EnumEditorService } from './enum-editor.service';
import { FarrisCommonModule } from '@farris/ui-common';
import { PerfectScrollbarModule } from '@farris/ui-perfect-scrollbar';
import { LookupModule } from '@farris/ui-lookup';
import { EnumEditorFooterComponent } from './enum-editor-footer.component';
var EnumEditorModule = /** @class */ (function () {
    function EnumEditorModule() {
    }
    /**
     * @return {?}
     */
    EnumEditorModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: EnumEditorModule,
            providers: [
                EnumEditorService
            ]
        };
    };
    EnumEditorModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        EnumEditorComponent,
                        EnumEditorInputComponent,
                        EnumEditorFooterComponent
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        InputGroupModule,
                        FarrisCommonModule,
                        MessagerModule.forRoot(),
                        PerfectScrollbarModule,
                        LookupModule,
                        LocaleModule
                    ],
                    entryComponents: [
                        EnumEditorComponent
                    ],
                    exports: [EnumEditorComponent, EnumEditorInputComponent]
                },] }
    ];
    return EnumEditorModule;
}());
export { EnumEditorModule };
