// @qdp/formular/esm5/lib/directive/help-before-event.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directive/help-before-event.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input } from '@angular/core';
import { LookupGridComponent } from '@farris/ui-lookup';
import { RestfulService } from '../service/restful.service';
import { of } from 'rxjs';
import { tap, switchMap } from 'rxjs/operators';
var HelpBeforeEventDirective = /** @class */ (function () {
    function HelpBeforeEventDirective(lookup, restfulService) {
        var _this = this;
        this.lookup = lookup;
        this.restfulService = restfulService;
        this.helpData = {};
        this.localHost = {};
        this.dynamicInput = [];
        this.dictPicking = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var options = {
                headers: {
                    'Content-Type': 'application/json'
                },
                responseType: 'text'
            };
            /** @type {?} */
            var paramvalues = {};
            _this.dynamicInput.forEach((/**
             * @param {?} currentInput
             * @return {?}
             */
            function (currentInput) {
                paramvalues[currentInput.paramcode] = currentInput.paramvalue;
            }));
            /** @type {?} */
            var params = {
                helpId: _this.helpData.fieldConfig.helpId,
                filterInfo: _this.helpData.fieldConfig.filterInfo,
                parameters: JSON.stringify(paramvalues)
            };
            /** @type {?} */
            var getHelpDataUrl = _this.localHost + '/api/runtime/bcc/v1.0/qdpsearch/getHelpData';
            /** @type {?} */
            var r$ = of({
                showDialog: true
            });
            return _this.restfulService.post(getHelpDataUrl, params, {}, options).pipe(tap((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                _this.lookup.items = JSON.parse(data);
            })), switchMap((/**
             * @return {?}
             */
            function () { return r$; })));
        });
    }
    /**
     * @return {?}
     */
    HelpBeforeEventDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        console.log(this.helpData, this.lookup);
        this.lookup.dictPicking = this.dictPicking;
        this.lookup.dictPicked = (/**
         * @param {?} rowData
         * @return {?}
         */
        function (rowData) {
            _this.helpData.paramvalue = rowData[_this.helpData.fieldConfig.valueField] + '|' + rowData[_this.helpData.fieldConfig.textField];
            /** @type {?} */
            var r$ = of({
                closeDialog: true
            });
            return r$;
        });
    };
    HelpBeforeEventDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[libHelpBeforeEvent]'
                },] }
    ];
    /** @nocollapse */
    HelpBeforeEventDirective.ctorParameters = function () { return [
        { type: LookupGridComponent },
        { type: RestfulService }
    ]; };
    HelpBeforeEventDirective.propDecorators = {
        helpData: [{ type: Input, args: ["libHelpBeforeEvent",] }],
        localHost: [{ type: Input }],
        dynamicInput: [{ type: Input }]
    };
    return HelpBeforeEventDirective;
}());
export { HelpBeforeEventDirective };
if (false) {
    /** @type {?} */
    HelpBeforeEventDirective.prototype.helpData;
    /** @type {?} */
    HelpBeforeEventDirective.prototype.localHost;
    /** @type {?} */
    HelpBeforeEventDirective.prototype.dynamicInput;
    /** @type {?} */
    HelpBeforeEventDirective.prototype.dictPicking;
    /**
     * @type {?}
     * @private
     */
    HelpBeforeEventDirective.prototype.lookup;
    /**
     * @type {?}
     * @private
     */
    HelpBeforeEventDirective.prototype.restfulService;
}
