// @farris/kendo-binding/esm2015/lib/directives/farris-datagrid/custom-size.directive.js
import { ChangeDetectorRef, Directive, ElementRef, Injector, Input } from '@angular/core';
import { DatagridComponent } from '@farris/ui-datagrid';
import { FrameContext } from '@farris/devkit';
export class CustomDatagridSizeDirective {
    constructor(dg, injector, cd, el, frameContext) {
        this.dg = dg;
        this.injector = injector;
        this.cd = cd;
        this.el = el;
        this.frameContext = frameContext;
        this.useCustomSize = false;
        this.sizeApiUri = '';
    }
    restService() {
        return this.frameContext.repository.restService;
    }
    ngOnInit() {
    }
    ngOnChanges(changes) {
        if (changes.sizeApiUri && !changes.sizeApiUri.isFirstChange()) {
            this.applyNewSize();
        }
    }
    ngAfterViewInit() {
        this.applyNewSize();
    }
    applyNewSize() {
        if (this.useCustomSize && this.sizeApiUri) {
            this.getSizeConfig().subscribe((cfg) => {
                if (this.dg && typeof this.dg.applyNewSize === 'function') {
                    this.dg.applyNewSize(cfg);
                }
            });
        }
    }
    getSizeConfig() {
        return this.restService().request(this.sizeApiUri, 'get');
    }
}
CustomDatagridSizeDirective.decorators = [
    { type: Directive, args: [{ selector: '[custom-size]' },] }
];
/** @nocollapse */
CustomDatagridSizeDirective.ctorParameters = () => [
    { type: DatagridComponent },
    { type: Injector },
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: FrameContext }
];
CustomDatagridSizeDirective.propDecorators = {
    useCustomSize: [{ type: Input, args: ['custom-size',] }],
    sizeApiUri: [{ type: Input }]
};
