// @farris/kendo-binding/esm2015/lib/directives/farris-datagrid/remote-summary.directive.js
import { Directive, Input } from '@angular/core';
import { ViewModel } from '@farris/devkit';
import { DatagridComponent } from '@farris/ui-datagrid';
export class FarrisDataGridRemoteSummaryDirective {
    constructor(viewModel, grid) {
        this.viewModel = viewModel;
        this.grid = grid;
        const appContext = this.viewModel.frameContext.getFormAppContext();
        if (appContext && appContext.hasOwnProperty('messagePipe')) {
            appContext.messagePipe.subscribe((message) => {
                const messageType = message && message.type || null;
                // 查询事件
                if (messageType === 'query') {
                    if (this.remoteSummaryCommand && this.grid.footerDataFrom === 'server') {
                        const paths = this.remoteSummaryCommand.split('.');
                        let func = null;
                        paths.forEach(path => {
                            func = func && func[path] || this[path];
                        });
                        if (typeof func === 'function') {
                            func().subscribe(result => {
                                this.grid.footerData = result;
                            });
                        }
                    }
                }
            });
        }
    }
    /**
     * 执行命令
     * @param commandName 命令名称
     * @param frameId frameId
     */
    execute(commandName, frameId) {
        if (!commandName || commandName === '' || commandName === 'undefined') {
            return;
        }
        let viewModel = this.viewModel;
        if (frameId) {
            const targetContext = viewModel.frameContext.appContext.frameContextManager.getFrameContextById(frameId);
            viewModel = targetContext.viewModel;
        }
        return viewModel[commandName]();
    }
}
FarrisDataGridRemoteSummaryDirective.decorators = [
    { type: Directive, args: [{
                selector: "[farris-remote-summary-command]"
            },] }
];
/** @nocollapse */
FarrisDataGridRemoteSummaryDirective.ctorParameters = () => [
    { type: ViewModel },
    { type: DatagridComponent }
];
FarrisDataGridRemoteSummaryDirective.propDecorators = {
    remoteSummaryCommand: [{ type: Input, args: ['farris-remote-summary-command',] }]
};
