// @gsp-svc/data-entity-adapter/fesm5/gsp-svc-data-entity-adapter.js
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { Injectable, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * data-entity-service对外提供服务类
 */
var DataEntityService = /** @class */ (function () {
    function DataEntityService(http) {
        this.http = http;
    }
    /**
     * @description 根据Id获取Schema
     * @param id be或vo的ID
     * @returns Schema
     */
    /**
     * \@description 根据Id获取Schema
     * @param {?} id be或vo的ID
     * @return {?} Schema
     */
    DataEntityService.prototype.getSchemaById = /**
     * \@description 根据Id获取Schema
     * @param {?} id be或vo的ID
     * @return {?} Schema
     */
    function (id) {
        return this.http.get("/api/runtime/sys/v1.0/dataentity/schema/id/" + id);
    };
    /**
     * 根据对象获取
     * @param obj 对象
     * @returns
     */
    /**
     * 根据对象获取
     * @param {?} map
     * @return {?}
     */
    DataEntityService.prototype.getSchemaByObj = /**
     * 根据对象获取
     * @param {?} map
     * @return {?}
     */
    function (map) {
        return this.http.post("/api/runtime/sys/v1.0/dataentity/schema/obj", map);
    };
    /**
     * 获取BE/VO
     * @param BE/VO的id
     */
    /**
     * 获取BE/VO
     * @param {?} id
     * @return {?}
     */
    DataEntityService.prototype.getBEOrVo = /**
     * 获取BE/VO
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http.get("/api/runtime/sys/v1.0/dataentity/schema/bevo/" + id);
    };
    DataEntityService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DataEntityService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return DataEntityService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var DataEntityModule = /** @class */ (function () {
    function DataEntityModule() {
    }
    DataEntityModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [
                        HttpClientModule
                    ],
                    entryComponents: [],
                    exports: [],
                    providers: [DataEntityService, HttpClient]
                },] }
    ];
    return DataEntityModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { DataEntityModule, DataEntityService };
