// @farris/kendo-binding/esm5/lib/pipes/formatPrimeNgTreeTableCellData.js
import { LocaleService } from '@farris/ui-locale';
import { Pipe, Injector, Optional } from '@angular/core';
import { ColumnFormatService } from '@farris/ui-common/column';
var FormatPrimeNgTreeTableCellData = /** @class */ (function () {
    function FormatPrimeNgTreeTableCellData(cfs, injector) {
        this.cfs = cfs;
        this.injector = injector;
        if (this.injector) {
            this.localeService = this.injector.get(LocaleService);
        }
    }
    FormatPrimeNgTreeTableCellData.prototype.transform = function (col, rowData) {
        var value = '';
        if (rowData && col && col.field) {
            value = this.getValue(col.field, rowData);
            var formatterFn = col.formatter;
            if (!formatterFn) {
                if (col.isMultilingualField) {
                    return this.getMultiLanguageValue(value);
                }
                else {
                    return value;
                }
            }
            else {
                return this.cfs.format(value, rowData, formatterFn);
            }
        }
        return '';
    };
    // 获取多语数据
    FormatPrimeNgTreeTableCellData.prototype.getMultiLanguageValue = function (valObj) {
        if (valObj && typeof valObj === 'object' && Object.keys(valObj).length > 0) {
            var localeId = this.localeService && this.localeService.localeId || 'zh-CHS';
            var value = valObj[localeId];
            return value === undefined ? '' : value;
        }
        else {
            return '';
        }
    };
    FormatPrimeNgTreeTableCellData.prototype.getValue = function (path, target) {
        if (!target) {
            return '';
        }
        var resultVal = '';
        if (path.indexOf('.') === -1) {
            resultVal = target[path];
        }
        else {
            resultVal = path.split('.').reduce((function (obj, key) {
                return obj && obj[key] || null;
            }), target);
        }
        return resultVal;
    };
    ;
    /**
     * 是否为帮助列
     * @param col col
     */
    FormatPrimeNgTreeTableCellData.prototype.isHelpColumn = function (col) {
        return col && col.editor && col.editor.options && col.editor.options.type === 'EditorTypes.LOOKUP' || false;
    };
    FormatPrimeNgTreeTableCellData.prototype.getHelpColumnValue = function (col, data) {
        var textField = this.getHelpControlTextField(col);
        var mapFields = this.getHelpControlMapFields(col);
        if (textField && mapFields && Object.keys(mapFields).length > 0) {
            var path = mapFields[textField];
            return this.getValue(path, data);
        }
        return '';
    };
    FormatPrimeNgTreeTableCellData.prototype.getHelpControlTextField = function (col) {
        return col && col.editor && col.editor.options && col.editor.options.textField || '';
    };
    FormatPrimeNgTreeTableCellData.prototype.getHelpControlMapFields = function (col) {
        return col && col.editor && col.editor.options && col.editor.options.mapFields || {};
    };
    FormatPrimeNgTreeTableCellData.decorators = [
        { type: Pipe, args: [{ name: 'formatPrimeNgTreeTableCellData', pure: false },] }
    ];
    /** @nocollapse */
    FormatPrimeNgTreeTableCellData.ctorParameters = function () { return [
        { type: ColumnFormatService },
        { type: Injector, decorators: [{ type: Optional }] }
    ]; };
    return FormatPrimeNgTreeTableCellData;
}());
export { FormatPrimeNgTreeTableCellData };
