// @qdp/common/esm5/lib/dnd-list/services/dnd-draggable.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
var DndDraggableService = /** @class */ (function () {
    function DndDraggableService() {
        this._isDragging = false;
        this._itemType = '';
        this._removeOnDrop = false;
        this.dropEndSource = new Subject();
    }
    /**
     * @param {?} state
     * @return {?}
     */
    DndDraggableService.prototype.setDraggingState = /**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this._isDragging = state;
    };
    /**
     * @return {?}
     */
    DndDraggableService.prototype.getDraggingState = /**
     * @return {?}
     */
    function () {
        return this._isDragging;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    DndDraggableService.prototype.setItemType = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this._itemType = value.toLowerCase();
    };
    /**
     * @return {?}
     */
    DndDraggableService.prototype.getItemType = /**
     * @return {?}
     */
    function () {
        return this._itemType;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    DndDraggableService.prototype.setDropCallback = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this._dropCallback = value;
    };
    /**
     * @return {?}
     */
    DndDraggableService.prototype.getDropCallback = /**
     * @return {?}
     */
    function () {
        return this._dropCallback;
    };
    /**
     * @param {?} dropEffect
     * @return {?}
     */
    DndDraggableService.prototype.setDropEffect = /**
     * @param {?} dropEffect
     * @return {?}
     */
    function (dropEffect) {
        this._dropEffect = dropEffect;
    };
    /**
     * @return {?}
     */
    DndDraggableService.prototype.getDropEffect = /**
     * @return {?}
     */
    function () {
        return this._dropEffect;
    };
    /**
     * @param {?} effectAllowed
     * @return {?}
     */
    DndDraggableService.prototype.setEffectAllowed = /**
     * @param {?} effectAllowed
     * @return {?}
     */
    function (effectAllowed) {
        this._effectAllowed = effectAllowed;
    };
    /**
     * @return {?}
     */
    DndDraggableService.prototype.getEffectAllowed = /**
     * @return {?}
     */
    function () {
        return this._effectAllowed;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    DndDraggableService.prototype.setRemoveOnDrop = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this._removeOnDrop = value;
    };
    /**
     * @return {?}
     */
    DndDraggableService.prototype.getRemoveOnDrop = /**
     * @return {?}
     */
    function () {
        return this._removeOnDrop;
    };
    DndDraggableService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DndDraggableService.ctorParameters = function () { return []; };
    return DndDraggableService;
}());
export { DndDraggableService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DndDraggableService.prototype._isDragging;
    /**
     * @type {?}
     * @private
     */
    DndDraggableService.prototype._itemType;
    /**
     * @type {?}
     * @private
     */
    DndDraggableService.prototype._dropCallback;
    /**
     * @type {?}
     * @private
     */
    DndDraggableService.prototype._dropEffect;
    /**
     * @type {?}
     * @private
     */
    DndDraggableService.prototype._effectAllowed;
    /**
     * @type {?}
     * @private
     */
    DndDraggableService.prototype._removeOnDrop;
    /** @type {?} */
    DndDraggableService.prototype.dropEndSource;
}
