// @gsp-wf/task-action-impl/esm5/lib/task-action-impl.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
var TaskActionImplService = /** @class */ (function () {
    function TaskActionImplService(injector, http) {
        this.injector = injector;
        this.http = http;
    }
    /**
     * 获取任务中心配置项
     * @param code 配置项的Code
     */
    /**
     * 获取任务中心配置项
     * @param {?} code 配置项的Code
     * @param {?=} ownerType
     * @param {?=} ownerId
     * @return {?}
     */
    TaskActionImplService.prototype.getTaskCenterConfigByCode = /**
     * 获取任务中心配置项
     * @param {?} code 配置项的Code
     * @param {?=} ownerType
     * @param {?=} ownerId
     * @return {?}
     */
    function (code, ownerType, ownerId) {
        /** @type {?} */
        var url = "/api/runtime/task/v1.0/task-center-setting/code?code=" + code;
        ownerType && (url += "&ownerType=" + ownerType);
        ownerId && (url += "&ownerId=" + ownerId);
        return this.http.get(url);
    };
    /**
      * 添加关注
      * @param sourceId sourceId
      */
    /**
     * 添加关注
     * @param {?} sourceId sourceId
     * @return {?}
     */
    TaskActionImplService.prototype.addFocus = /**
     * 添加关注
     * @param {?} sourceId sourceId
     * @return {?}
     */
    function (sourceId) {
        /** @type {?} */
        var url = "/api/runtime/task/v1.0/focus?sourceId=" + sourceId;
        return this.http.post(url, '');
    };
    /**
     * 移除关注
     * @param processInstanceId 流程实例id
     */
    /**
     * 移除关注
     * @param {?} processInstanceId 流程实例id
     * @return {?}
     */
    TaskActionImplService.prototype.removeFocus = /**
     * 移除关注
     * @param {?} processInstanceId 流程实例id
     * @return {?}
     */
    function (processInstanceId) {
        /** @type {?} */
        var url = "/api/runtime/task/v1.0/focus?processInstanceId=" + processInstanceId;
        return this.http.delete(url, '');
    };
    /**
     * 是否关注了该任务
     * @param processInstanceId 流程实例id
     */
    /**
     * 是否关注了该任务
     * @param {?} processInstanceId 流程实例id
     * @return {?}
     */
    TaskActionImplService.prototype.ifFocused = /**
     * 是否关注了该任务
     * @param {?} processInstanceId 流程实例id
     * @return {?}
     */
    function (processInstanceId) {
        /** @type {?} */
        var url = "/api/runtime/task/v1.0/focus/?processInstanceId=" + processInstanceId;
        return this.http.get(url);
    };
    TaskActionImplService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    TaskActionImplService.ctorParameters = function () { return [
        { type: Injector },
        { type: HttpService }
    ]; };
    return TaskActionImplService;
}());
export { TaskActionImplService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaskActionImplService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    TaskActionImplService.prototype.http;
}
