// @gsp-cmp/querysolution/esm5/lib/service/edit.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/edit.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
/** @type {?} */
var itemIndex = (/**
 * @param {?} item
 * @param {?} data
 * @return {?}
 */
function (item, data) {
    for (var idx = 0; idx < data.length; idx++) {
        if (data[idx].id === item.id) {
            return idx;
        }
    }
    return -1;
});
var ɵ0 = itemIndex;
var EditService = /** @class */ (function () {
    function EditService() {
        this.createdItems = [];
        this.updatedItems = [];
    }
    /**
     * @param {?} item
     * @return {?}
     */
    EditService.prototype.update = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        if (!this.isNew(item)) {
            /** @type {?} */
            var index = itemIndex(item, this.updatedItems);
            if (index !== -1) {
                this.updatedItems.splice(index, 1, item);
            }
            else {
                this.updatedItems.push(item);
            }
        }
        else {
            /** @type {?} */
            var index = this.createdItems.indexOf(item);
            this.createdItems.splice(index, 1, item);
        }
    };
    /**
     * @param {?} item
     * @return {?}
     */
    EditService.prototype.isNew = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        return !item.id;
    };
    /**
     * @param {?} target
     * @param {?} source
     * @return {?}
     */
    EditService.prototype.assignValues = /**
     * @param {?} target
     * @param {?} source
     * @return {?}
     */
    function (target, source) {
        Object.assign(target, source);
    };
    EditService.decorators = [
        { type: Injectable }
    ];
    return EditService;
}());
export { EditService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    EditService.prototype.createdItems;
    /**
     * @type {?}
     * @private
     */
    EditService.prototype.updatedItems;
}
export { ɵ0 };
