// @gsp-wf/rtdevkit/esm5/lib/runtime/components/task-transfer/task-transfer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { TaskService } from '../../services/task.service';
import { UserService } from '../../services/user.service';
import { UserHelperComponent } from '../user-helper/user-helper.component';
import { NotifyService } from '@farris/ui-notify';
import { I18nService } from '../../services/i18n.service';
var TaskTransferComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaskTransferComponent, _super);
    function TaskTransferComponent(taskSvc, userSvc, notify, i18nService) {
        var _this = _super.call(this, i18nService, userSvc) || this;
        _this.taskSvc = taskSvc;
        _this.userSvc = userSvc;
        _this.notify = notify;
        _this.i18nService = i18nService;
        _this.completed = new EventEmitter();
        _this.singleSelect = true;
        _this.paginationOptions.id = 'WF-TaskTransfer-DataTable-Pagination';
        return _this;
    }
    Object.defineProperty(TaskTransferComponent.prototype, "transfer_context", {
        set: /**
         * @param {?} context
         * @return {?}
         */
        function (context) {
            this.transferContext = tslib_1.__assign({}, context, { assignee: '' });
            this.selectedUsers = [];
            this.showModal(!!this.transferContext.workItemId);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TaskTransferComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    TaskTransferComponent.prototype.handleOk = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.selectedUsers = this.userTable.selections;
        // 需要校验选中唯一用户
        if (!this.selectedUsers) {
            this.notify.info(this.i18nService.getValue('static.transfer.selectHandover'));
            return;
        }
        this.transferContext.assignee = this.selectedUsers['ID'];
        this.taskSvc.transfer(this.transferContext).subscribe((/**
         * @return {?}
         */
        function () {
            _this.notify.success(_this.i18nService.getValue('static.transfer.successTransfer'));
            _this.completed.emit(true);
            _this.showModal(false);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            if (error && error.error && error.error.Message) {
                _this.notify.error(error.error.Message);
            }
        }));
    };
    TaskTransferComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-task-transfer',
                    template: "<farris-dialog #allUserDialog [buttons]=\"buttonRef\" [width]=\"850\" [height]=\"520\" [title]=\"'\u7CFB\u7EDF\u7528\u6237'\">\r\n  <div class=\"d-flex flex-column my-data-grid h-100\" style=\"padding: 0 14px;\">\r\n    <div class=\"flex-auto\" style=\"position:relative;\">\r\n      <farris-datatable #allUserTable [columns]=\"columns\" [data]=\"userListData | paginate: paginationOptions\"\r\n        [showFilterBar]='true' [singleSelect]=\"singleSelect\" [idField]=\"'ID'\" [remote]=\"'server'\" (search)=\"search($event)\"\r\n        [pagination]='false'></farris-datatable>\r\n    </div>\r\n    <div style=\"position:relative;\">\r\n      <pagination-controls [id]=\"paginationOptions.id\" [previousLabel]=\"' '\" [nextLabel]=\"' '\"\r\n        (pageChange)=\"onPageChange($event)\" (pageSizeChange)=\"onPageSizeChange($event)\">\r\n      </pagination-controls>\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n<ng-template #buttonRef>\r\n  <button class=\"btn btn-sm btn-primary px-3\" (click)=\"handleOk()\">\u786E\u5B9A</button>\r\n  <button class=\"btn btn-sm btn-default px-3\" (click)=\"handleCancel()\">\u53D6\u6D88</button>\r\n</ng-template>",
                    styles: [".my-data-grid .filter-row .search-fields{padding-left:0!important}.my-data-grid .filter-row .search-text{padding-right:0!important}"]
                }] }
    ];
    /** @nocollapse */
    TaskTransferComponent.ctorParameters = function () { return [
        { type: TaskService },
        { type: UserService },
        { type: NotifyService },
        { type: I18nService }
    ]; };
    TaskTransferComponent.propDecorators = {
        transfer_context: [{ type: Input }],
        completed: [{ type: Output }]
    };
    return TaskTransferComponent;
}(UserHelperComponent));
export { TaskTransferComponent };
if (false) {
    /** @type {?} */
    TaskTransferComponent.prototype.completed;
    /** @type {?} */
    TaskTransferComponent.prototype.transferContext;
    /**
     * @type {?}
     * @private
     */
    TaskTransferComponent.prototype.taskSvc;
    /**
     * @type {?}
     * @protected
     */
    TaskTransferComponent.prototype.userSvc;
    /**
     * @type {?}
     * @protected
     */
    TaskTransferComponent.prototype.notify;
    /**
     * @type {?}
     * @protected
     */
    TaskTransferComponent.prototype.i18nService;
}
