// @farris/ui-lookup/esm5/lib/lookup-displaytype.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 收藏列字段名称
 * @type {?}
 */
export var FAVORITE_FIELD_NAME = '__favorite__';
/** @type {?} */
export var INPUT_GROUP_ICON = '<i class="f-icon f-icon-lookup"></i>';
/** @enum {string} */
var LookupGridDisplayType = {
    List: 'LIST',
    TreeList: 'TREELIST',
    NavList: 'NAVLIST',
    NavTreeList: 'NAVTREELIST',
    NavListTree: 'NAVLISTTREE',
};
export { LookupGridDisplayType };
/** @enum {string} */
var FavoriteIcon = {
    /** 已收藏 */
    yes: '<span class="f-icon f-icon-star f-lookup-favorite" style="font-size:12px" ></span>',
    /** 未收藏 */
    no: '<span class="f-icon f-icon-star-outline f-lookup-favorite" style="font-size:12px"></span>',
    /** 移除收藏 */
    delete: '<span class="f-icon f-icon-star f-lookup-unfavorite" style="font-size:12px"></span>',
    /** 移除已选记录 */
    remove: '<span class="f-icon f-icon-minus-circle" style="font-size:12px; color: #dd2438;cursor: pointer"></span>',
};
export { FavoriteIcon };
/** @enum {string} */
var FavoriteAction = {
    /** 添加收藏 */
    add: 'append item to favorite.',
    /** 移除收藏 */
    delete: 'remove favorite.',
};
export { FavoriteAction };
/**
 * @record
 */
export function ExtInfoFormatterParam() { }
if (false) {
    /** @type {?|undefined} */
    ExtInfoFormatterParam.prototype.bindingData;
    /** @type {?|undefined} */
    ExtInfoFormatterParam.prototype.instance;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * @record
 */
export function CustomDataFormatter() { }
if (false) {
    /**
     * 行样式
     * @type {?|undefined}
     */
    CustomDataFormatter.prototype.rowStyler;
    /**
     * 单元格样式
     * @type {?|undefined}
     */
    CustomDataFormatter.prototype.cellStyler;
    /* Skipping unhandled member: [fieldName: string]: DataFormatter ;*/
}
/**
 * @record
 */
export function QuickSelectOption() { }
if (false) {
    /**
     * 是否启用
     * @type {?|undefined}
     */
    QuickSelectOption.prototype.enable;
    /**
     * 显示top 记录数，默认 10
     * @type {?|undefined}
     */
    QuickSelectOption.prototype.showItemsCount;
    /**
     * 显示文本字段
     * @type {?|undefined}
     */
    QuickSelectOption.prototype.textField;
    /**
     * 自定义格式化
     * @type {?|undefined}
     */
    QuickSelectOption.prototype.formatter;
    /** @type {?|undefined} */
    QuickSelectOption.prototype.footerHeight;
    /**
     * 显示搜索更新 按钮
     * @type {?|undefined}
     */
    QuickSelectOption.prototype.showMore;
    /**
     * 最小宽度, 默认 200px
     * @type {?|undefined}
     */
    QuickSelectOption.prototype.minWidth;
    /**
     * 启用回退键移除选中项，默认为 false
     * @type {?|undefined}
     */
    QuickSelectOption.prototype.enableBackspace;
}
/** @type {?} */
export var QuickSelectDefaultOptions = {
    enable: false,
    showMore: true,
    showItemsCount: 10,
    footerHeight: 0,
    minWidth: 200,
    enableBackspace: false
}
/**
 * - both: 全部显示
 * - onlyfield: 仅显示字段
 * - onlyinput: 仅显示搜索输入框 */
;
/** @enum {string} */
var SearchBarMode = {
    /** 显示字段查询与搜索输入框 */
    both: 'both',
    /** 仅显示字段 */
    onlyinput: 'onlyinput',
    /** 仅显示搜索输入框 */
    onlyfield: 'onlyfield',
};
export { SearchBarMode };
