// @qdp/ide-cmp/esm5/lib/service/paras-mapping/paras-mapping.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { ParasMappingManagerService } from '@qdp/common';
var ParasMappingService = /** @class */ (function () {
    function ParasMappingService(parasMappingManager) {
        var _this = this;
        this.parasMappingManager = parasMappingManager;
        this.paramMappingInfo = new Subject();
        this.parasMappingManager.parasmappingInfoChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.paramMappingInfo.next(value);
        }));
    }
    /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    ParasMappingService.prototype.getParasMapping = /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (id, queryId, queryRelativeUrl) {
        return this.parasMappingManager.getParasMapping(id, queryId, queryRelativeUrl);
    };
    /**
     * @param {?} paramMappingInfo
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    ParasMappingService.prototype.saveParasMapping = /**
     * @param {?} paramMappingInfo
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (paramMappingInfo, queryRelativeUrl) {
        return this.parasMappingManager.saveParasMapping(paramMappingInfo, queryRelativeUrl);
    };
    /**
     * @param {?} datasourceID
     * @param {?} queryID
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    ParasMappingService.prototype.deleteParasMapping = /**
     * @param {?} datasourceID
     * @param {?} queryID
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (datasourceID, queryID, queryRelativeUrl) {
        return this.parasMappingManager.deleteParasMapping(datasourceID, queryID, queryRelativeUrl);
    };
    ParasMappingService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ParasMappingService.ctorParameters = function () { return [
        { type: ParasMappingManagerService }
    ]; };
    return ParasMappingService;
}());
export { ParasMappingService };
if (false) {
    /** @type {?} */
    ParasMappingService.prototype.paramMappingInfo;
    /**
     * @type {?}
     * @private
     */
    ParasMappingService.prototype.parasMappingManager;
}
