// @gsp-cmp/ubcc-param-vo/esm2015/lib/vo/serializer/template-assignment-serializer/dimension-info-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/template-assignment-serializer/dimension-info-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DimensionInfo } from '../../param-template-assignment-vo/dimension-info';
import { DimensionSerializer } from '../dimension-serializer/dimension-serializer';
import { DimensionDataSerializer } from './dimension-data-serializer';
export class DimensionInfoSerializer {
    /**
     * @param {?} dimensionInfo
     * @return {?}
     */
    serialize(dimensionInfo) {
        if (!dimensionInfo) {
            return null;
        }
        /** @type {?} */
        let jobj = new Object();
        jobj['dataModificationAllow'] = dimensionInfo.dataModificationAllow;
        if (dimensionInfo.data) {
            jobj['data'] = dimensionInfo.data.map((/**
             * @param {?} dimensionData
             * @return {?}
             */
            dimensionData => new DimensionDataSerializer().serialize(dimensionData)));
        }
        if (dimensionInfo.definition) {
            /** @type {?} */
            let serializer = new DimensionSerializer();
            jobj['definition'] = serializer.serialize(dimensionInfo.definition);
        }
        return jobj;
    }
    /**
     * @param {?} jobj
     * @return {?}
     */
    deserialize(jobj) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        let dimensionInfo = new DimensionInfo();
        dimensionInfo.dataModificationAllow = (/** @type {?} */ (jobj['dataModificationAllow']));
        if (jobj['data']) {
            dimensionInfo.data = jobj['data'].map((/**
             * @param {?} dimensionData
             * @return {?}
             */
            dimensionData => new DimensionDataSerializer().deserialize(dimensionData)));
        }
        if (jobj['definition']) {
            /** @type {?} */
            let serializer = new DimensionSerializer();
            dimensionInfo.definition = serializer.deserialize(jobj['definition']);
        }
        return dimensionInfo;
    }
}
