// @farris/ui-datagrid/esm2015/lib/components/auto-height/auto-height.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DatagridComponent } from './../../datagrid.component';
import { Component, Injector, Input } from '@angular/core';
/**
 * 自动高度
 * 禁用分页，禁用固定表头，禁用虚拟加载，禁用固定列
 * 支持单元格编辑, 支持排序, 支持拖动列宽
 */
export class AutoHeightComponent {
    /**
     * @param {?} dg
     * @param {?} inject
     */
    constructor(dg, inject) {
        this.dg = dg;
        this.inject = inject;
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
}
AutoHeightComponent.decorators = [
    { type: Component, args: [{
                selector: 'auto-height',
                template: "<table style=\"width:100%\" class=\"f-datagrid-ah\" cellpadding=\"0\" cellspacing=\"0\" border=\"0\">\r\n    <colgroup>\r\n        <col *ngIf=\"dg.showLineNumber\" [width]=\"dg.lineNumberWidth + 'px'\">\r\n        <ng-container *ngFor=\"let col of columnsGroup?.normalColumns\">\r\n            <col [width]=\"col.width + 'px'\">\r\n        </ng-container>\r\n    </colgroup>\r\n    <thead>\r\n        <tr class=\"f-datagrid-header-row\" *ngFor=\"let cols of columns\">\r\n            <td class=\"f-datagrid-header-cell\" *ngIf=\"dg.showLineNumber\"></td>\r\n            <ng-container *ngFor=\"let col of cols\">\r\n                <ng-container *ngIf=\"col.colspan && !col.fixed\">\r\n                    <td class=\"f-datagrid-header-cell f-datagrid-header-cell-group\" \r\n                        [attr.align]=\"col.halign\"\r\n                        [attr.colspan]=\"col.colspan\">\r\n                        <div class=\"f-datagrid-header-cell-content\" >\r\n                            {{ col.title }}\r\n                        </div>\r\n                    </td>\r\n                </ng-container>\r\n                <ng-container *ngIf=\"!col.colspan && !col.fixed\">\r\n                    <td class=\"f-datagrid-header-cell\" \r\n                        [attr.align]=\"col.halign\" \r\n                        [attr.field]=\"col.field\"\r\n                        [attr.rowspan]=\"col.rowspan\">\r\n                        <div class=\"f-datagrid-header-cell-content\">\r\n                            {{ col.title }}\r\n                        </div>\r\n                    </td>\r\n                </ng-container>\r\n            </ng-container>\r\n        </tr>\r\n    </thead>\r\n    <tbody>\r\n        <tr class=\"f-datagrid-body-row\" [attr.id]=\"row | rowDataId\" *ngFor=\"let row of data; trackBy: dg.trackByRows; let i = index\">\r\n            <td class=\"f-datagrid-cell f-datagrid-cell-rownumber\" [style.width.px]=\"dg.lineNumberWidth\" *ngIf=\"dg.showLineNumber\">{{ i+1 }}</td>\r\n            <td class=\"f-datagrid-cell\" [attr.align]=\"col.align\" *ngFor=\"let col of columnsGroup?.normalColumns; let ci = index;\">\r\n                <span [innerHtml]=\" col | formatCellData: row | safe: 'html'\"></span>\r\n            </td>\r\n        </tr>\r\n        <ng-container *ngIf=\"dg.showFooter && data && data.length\">\r\n            <tr class=\"f-datagrid-body-row\" *ngFor=\"let row of dg.footerData\">\r\n                <td class=\"f-datagrid-cell f-datagrid-cell-rownumber\" [style.width.px]=\"dg.lineNumberWidth\" *ngIf=\"dg.showLineNumber\"></td>\r\n                <td class=\"f-datagrid-cell\" [attr.align]=\"col.align\" *ngFor=\"let col of columnsGroup?.normalColumns\" [attr.title]=\"col | formatCellData: row \">\r\n                    <div class=\"f-datagrid-cell-content\">\r\n                        <span *ngIf=\"col.footer && col.footer.formatter\" [innerHtml]=\" col | formatCellData: row : false: true | safe: 'html'\"></span>\r\n                        <span *ngIf=\"!col.footer || !col.footer.formatter\">{{ col | formatCellData: row: false: true }}</span>\r\n                    </div>\r\n                </td>\r\n            </tr>\r\n        </ng-container>\r\n    </tbody>\r\n</table>"
            }] }
];
/** @nocollapse */
AutoHeightComponent.ctorParameters = () => [
    { type: DatagridComponent },
    { type: Injector }
];
AutoHeightComponent.propDecorators = {
    columns: [{ type: Input }],
    columnsGroup: [{ type: Input }],
    data: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    AutoHeightComponent.prototype.columns;
    /** @type {?} */
    AutoHeightComponent.prototype.columnsGroup;
    /** @type {?} */
    AutoHeightComponent.prototype.data;
    /** @type {?} */
    AutoHeightComponent.prototype.dg;
    /**
     * @type {?}
     * @private
     */
    AutoHeightComponent.prototype.inject;
}
