// @gsp-aif/runtime-api/esm5/lib/aif-push-select-creation-rule/service/aif-push-select-creation-rule.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, ComponentFactoryResolver } from "@angular/core";
import { BsModalService } from '@farris/ui-modal';
import { AifPushSelectCreationRuleComponent } from '../aif-push-select-creation-rule/aif-push-select-creation-rule.component';
/**
 * 供外部调用的推式选择生单规则服务类接口
 */
var AifPushSelectCreationRuleService = /** @class */ (function () {
    function AifPushSelectCreationRuleService(injector, cfr, modalService) {
        this.injector = injector;
        this.cfr = cfr;
        this.modalService = modalService;
        /**
         * 内置模态框配置
         */
        this.innerModalOptions = {
            title: '生单规则选择',
            width: 650,
            height: 415,
            buttons: null,
            showButtons: false,
        };
    }
    /**
     * 显示选择生单规则模态框，点击确定按钮时返回选中的生单规则
     * @param args 查询生单规则所需参数
     * @param modalOptions 模态框配置,可为空
     */
    /**
     * 显示选择生单规则模态框，点击确定按钮时返回选中的生单规则
     * @param {?} args 查询生单规则所需参数
     * @param {?=} modalOptions 模态框配置,可为空
     * @return {?}
     */
    AifPushSelectCreationRuleService.prototype.showCrRuleSelectModal = /**
     * 显示选择生单规则模态框，点击确定按钮时返回选中的生单规则
     * @param {?} args 查询生单规则所需参数
     * @param {?=} modalOptions 模态框配置,可为空
     * @return {?}
     */
    function (args, modalOptions) {
        /** @type {?} */
        var cmpF = this.cfr.resolveComponentFactory(AifPushSelectCreationRuleComponent);
        /** @type {?} */
        var cmpR = cmpF.create(this.injector);
        cmpR.instance.getCreationRule(args);
        cmpR.instance.bsModalRef = this.modalService.show(cmpR, this.isNullOrUndefined(modalOptions) ? this.innerModalOptions : modalOptions);
        return cmpR.instance.subject.asObservable();
    };
    /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    AifPushSelectCreationRuleService.prototype.isNullOrUndefined = /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        return arg == null;
    };
    AifPushSelectCreationRuleService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AifPushSelectCreationRuleService.ctorParameters = function () { return [
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: BsModalService }
    ]; };
    return AifPushSelectCreationRuleService;
}());
export { AifPushSelectCreationRuleService };
if (false) {
    /**
     * 内置模态框配置
     * @type {?}
     * @private
     */
    AifPushSelectCreationRuleService.prototype.innerModalOptions;
    /**
     * @type {?}
     * @private
     */
    AifPushSelectCreationRuleService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    AifPushSelectCreationRuleService.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    AifPushSelectCreationRuleService.prototype.modalService;
}
