// @farris/ui-tooltip/esm2015/lib/attention.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, RendererFactory2, NgZone } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function AttentionElement() { }
if (false) {
    /** @type {?} */
    AttentionElement.prototype.idOrEl;
    /** @type {?|undefined} */
    AttentionElement.prototype.type;
    /** @type {?|undefined} */
    AttentionElement.prototype.class;
}
export class AttentionService {
    /**
     * @param {?} renderFac
     * @param {?} ngZone
     */
    constructor(renderFac, ngZone) {
        this.ngZone = ngZone;
        this.renderer = renderFac.createRenderer(null, null);
    }
    /**
     * @param {?} els
     * @param {?=} timer
     * @return {?}
     */
    catchAttention(els, timer = 600) {
        if (els && els.length > 0) {
            /** @type {?} */
            let attentionEls = [];
            for (let k = 0; k < els.length; k++) {
                /** @type {?} */
                var attentionEl;
                if (typeof els[k].idOrEl == 'string') {
                    attentionEl = document.getElementById(els[k].idOrEl);
                }
                else {
                    attentionEl = els[k].idOrEl;
                }
                if (attentionEl) {
                    attentionEls.push(attentionEl);
                    this.renderer.addClass(attentionEl, 'f-catch-attention-debounce');
                }
            }
            // 移除class
            this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => {
                setTimeout((/**
                 * @return {?}
                 */
                () => {
                    for (let m = 0; m < attentionEls.length; m++) {
                        this.renderer.removeClass(attentionEls[m], 'f-catch-attention-debounce');
                    }
                }), timer);
            }));
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
AttentionService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
AttentionService.ctorParameters = () => [
    { type: RendererFactory2 },
    { type: NgZone }
];
/** @nocollapse */ AttentionService.ngInjectableDef = i0.defineInjectable({ factory: function AttentionService_Factory() { return new AttentionService(i0.inject(i0.RendererFactory2), i0.inject(i0.NgZone)); }, token: AttentionService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    AttentionService.prototype.renderer;
    /** @type {?} */
    AttentionService.prototype.ngZone;
}
