// @farris/ui-tooltip/esm2015/lib/inputTips.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Renderer2, NgZone, Input } from '@angular/core';
export class FInputTipsDirective {
    /**
     * @param {?} el
     * @param {?} ngzone
     * @param {?} render
     */
    constructor(el, ngzone, render) {
        this.el = el;
        this.ngzone = ngzone;
        this.render = render;
        this.cloneEl = null;
        this.tipsText = "";
        this._enableTips = true;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set enableTips(value) {
        if (this._enableTips !== value) {
            this._enableTips = value;
            this.changeEventBind();
        }
    }
    /**
     * @return {?}
     */
    get enableTips() {
        return this._enableTips;
    }
    /**
     * @private
     * @return {?}
     */
    createCloneEl() {
        if (!this.el.nativeElement) {
            return;
        }
        /** @type {?} */
        const global = {
            cloneCSSProperties: [
                'line-height', 'letter-spacing',
                'font-size', 'font-family', 'font-style', 'font-weight', 'border',
                'padding', 'height'
            ]
        };
        /** @type {?} */
        const el$ = window.getComputedStyle(this.el.nativeElement);
        this.cloneEl = this.render.createElement('span');
        this.cloneEl['innerText'] = el$['value'];
        global.cloneCSSProperties.forEach((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        (item, index) => {
            this.cloneEl.style[item] = el$.getPropertyValue(item);
        }));
        this.render.addClass(this.cloneEl, 'f-dir-hidden-el');
        this.render.appendChild(document.body, this.cloneEl);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.changeEventBind();
    }
    /**
     * @return {?}
     */
    changeEventBind() {
        if (this.enableTips) {
            this.ngzone.runOutsideAngular((/**
             * @return {?}
             */
            () => {
                this.mouseEnterListener = this.onMouseEnterHandler.bind(this);
                this.el.nativeElement.addEventListener('mouseenter', this.mouseEnterListener);
                this.mouseLeaveListener = this.onMouseLeaveHandler.bind(this);
                this.el.nativeElement.addEventListener('mouseleave', this.mouseLeaveListener);
            }));
        }
        else {
            if (this.mouseEnterListener) {
                this.el.nativeElement.removeEventListener('mouseenter', this.mouseEnterListener);
                this.mouseEnterListener = null;
                this.el.nativeElement.removeEventListener('mouseleave', this.mouseLeaveListener);
                this.mouseLeaveListener = null;
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.cloneEl) {
            this.render.removeChild(document.body, this.cloneEl);
        }
    }
    /**
     * @return {?}
     */
    onMouseEnterHandler() {
        this.changeTitle();
    }
    /**
     * @return {?}
     */
    onMouseLeaveHandler() {
        // 移除提示
        this.render.setAttribute(this.el.nativeElement, 'title', '');
    }
    /**
     * @private
     * @return {?}
     */
    changeTitle() {
        if (!this.cloneEl) {
            this.createCloneEl();
        }
        this.tipsText = this.el.nativeElement['value'];
        this.cloneEl['innerText'] = this.tipsText;
        this.ngzone.runOutsideAngular((/**
         * @return {?}
         */
        () => {
            setTimeout((/**
             * @return {?}
             */
            () => {
                if (this.cloneEl.getBoundingClientRect().width > this.el.nativeElement.getBoundingClientRect().width) {
                    this.render.setAttribute(this.el.nativeElement, "title", this.tipsText);
                }
                else {
                    this.onMouseLeaveHandler();
                }
            }), 0);
        }));
    }
}
FInputTipsDirective.decorators = [
    { type: Directive, args: [{
                selector: '[f-input-tips]'
            },] }
];
/** @nocollapse */
FInputTipsDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NgZone },
    { type: Renderer2 }
];
FInputTipsDirective.propDecorators = {
    enableTips: [{ type: Input, args: ['f-input-tips',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    FInputTipsDirective.prototype.cloneEl;
    /**
     * @type {?}
     * @private
     */
    FInputTipsDirective.prototype.tipsText;
    /** @type {?} */
    FInputTipsDirective.prototype.mouseEnterListener;
    /** @type {?} */
    FInputTipsDirective.prototype.mouseLeaveListener;
    /**
     * @type {?}
     * @private
     */
    FInputTipsDirective.prototype._enableTips;
    /**
     * @type {?}
     * @private
     */
    FInputTipsDirective.prototype.el;
    /** @type {?} */
    FInputTipsDirective.prototype.ngzone;
    /**
     * @type {?}
     * @private
     */
    FInputTipsDirective.prototype.render;
}
