// @farris/ui-dropdown/esm2015/lib/f-dropdown-toggle.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostBinding, HostListener, ElementRef, Input, } from "@angular/core";
import { FDropdownDirective } from "./f-dropdown.directive";
export class FDropdownToggleDirective {
    /**
     * @param {?} dropdown
     * @param {?} elementRef
     */
    constructor(dropdown, elementRef) {
        this.dropdown = dropdown;
        this.elementRef = elementRef;
        this._disabled = false;
        this.haspopup = true;
        this.nativeEle = null;
        dropdown.toggleElement = elementRef.nativeElement;
        this.nativeEle = elementRef.nativeElement;
        this.dropdown.getOpenState().subscribe((/**
         * @param {?} state
         * @return {?}
         */
        (state) => {
            if (this.dropdown.isSubDP && this.dropdown.isOpen) {
                if (this.nativeEle.className.indexOf("active") < 0) {
                    this.nativeEle.className += " active";
                }
            }
            else {
                if (this.nativeEle.className.indexOf("active") > -1) {
                    this.nativeEle.className = this.nativeEle.className.replace(" active", " ");
                }
            }
        }));
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set dpDisabled(value) {
        this._disabled = value;
    }
    /**
     * @return {?}
     */
    get disableCls() {
        return this._disabled;
    }
    /**
     * @param {?} ev
     * @return {?}
     */
    onclick(ev) {
        if (this._disabled) {
            return;
        }
        if (this.dropdown.isSubDP) {
            ev.stopImmediatePropagation();
        }
        this.dropdown.toggle();
    }
}
FDropdownToggleDirective.decorators = [
    { type: Directive, args: [{
                selector: "[fDropdownToggle]",
            },] }
];
/** @nocollapse */
FDropdownToggleDirective.ctorParameters = () => [
    { type: FDropdownDirective },
    { type: ElementRef }
];
FDropdownToggleDirective.propDecorators = {
    dpDisabled: [{ type: Input }],
    disableCls: [{ type: HostBinding, args: ["class.disabled",] }],
    haspopup: [{ type: HostBinding, args: ["attr.aria-haspopup",] }],
    onclick: [{ type: HostListener, args: ["click", ["$event"],] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    FDropdownToggleDirective.prototype._disabled;
    /** @type {?} */
    FDropdownToggleDirective.prototype.haspopup;
    /**
     * @type {?}
     * @private
     */
    FDropdownToggleDirective.prototype.nativeEle;
    /**
     * @type {?}
     * @private
     */
    FDropdownToggleDirective.prototype.dropdown;
    /**
     * @type {?}
     * @private
     */
    FDropdownToggleDirective.prototype.elementRef;
}
