// @farris/command-services/esm5/lib/rtf-message/message-pipe.js
import { UID } from '@farris/devkit';
var MessagePipe = /** @class */ (function () {
    function MessagePipe(token, rtfService) {
        this.token = token;
        this.rtfService = rtfService;
    }
    /**
     * 向管道发送消息
     * @param message 消息
     */
    MessagePipe.prototype.sendMessage = function (message) {
        this.rtfService.subjectNotify(this.token, message);
        return this;
    };
    /**
     * 监听管道消息
     * @param handler 消息处理器
     */
    MessagePipe.prototype.listen = function (handler) {
        var _this = this;
        var observerToken = UID.create();
        this.rtfService.subjectResponse(this.token, handler, observerToken);
        var unListen = function () {
            _this.rtfService.responseUnSubscribe(_this.token, observerToken);
        };
        return unListen;
    };
    /**
     * 销毁管道
     */
    MessagePipe.prototype.destory = function () {
        this.rtfService.subjectRemove(this.token);
    };
    return MessagePipe;
}());
export { MessagePipe };
