// @farris/ui-adminorganization-selector/esm2015/utils/api.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/api.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { of } from 'rxjs';
export class ApiService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * 获取行政组织或系统用户
     * @param {?} param
     * @return {?} observable
     */
    getOrganization(param) {
        /** @type {?} */
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        return this.http.post(this.absUrl + '/org', param, options);
    }
    /**
     * 获取行政组织（全加载）
     * @param {?} param
     * @return {?} observable
     */
    getOrganizations(param) {
        /** @type {?} */
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        /** @type {?} */
        let url = this.getOrgsUrl + `/getorgs?withDisabledOrg=${param.withDisabledOrg}&gettingCurrentOrg=${param.gettingCurrentOrg}&onlyDisplayOrg=${param.onlyDisplayOrg}`;
        return this.http.post(url, options);
    }
    /**
     * 获取行政组织（分层加载）
     * @param {?} param
     * @return {?} observable
     */
    getOrganizationsWithLayer(param) {
        /** @type {?} */
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        /** @type {?} */
        let url = this.getOrgsUrl + `/getOrgsWithLayer?withDisabledOrg=${param.withDisabledOrg}`;
        if (param.parentOrgId) {
            url = url + `&parentOrgId=${param.parentOrgId}`;
        }
        url = url + `&gettingCurrentOrg=${param.gettingCurrentOrg}&onlyDisplayOrg=${param.onlyDisplayOrg}`;
        return this.http.post(url, options);
    }
    /**
     * 根据id获取行政组织
     * @param {?} ids
     * @return {?} observable
     */
    getOrganizationsWithIds(ids) {
        /** @type {?} */
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        return this.http.post(this.getOrgsUrl + '/getorgsbyids', JSON.stringify(ids), options);
    }
    /**
     * 根据名称获取行政组织
     * @param {?} queryFilter
     * @return {?} observable
     */
    getOrganizationsWithFilter(queryFilter) {
        /** @type {?} */
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        return this.http.post(this.getOrgsUrl + '/query', JSON.stringify(queryFilter), options);
    }
    /**
     * @param {?} absOrgType
     * @return {?}
     */
    getOrgGroups(absOrgType) {
        /** @type {?} */
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        return this.http.get(this.groupApiUrl + '/' + absOrgType, options);
    }
    /**
     * @param {?} groupId
     * @param {?} absOrgType
     * @return {?}
     */
    getGroupMembers(groupId, absOrgType) {
        /** @type {?} */
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        if (!groupId) {
            return of(null);
        }
        /** @type {?} */
        const url = this.groupApiUrl + '/' + groupId + '/absOrgType/' + absOrgType;
        return this.http.get(url, options);
    }
    /**
     * @param {?} groupInfo
     * @return {?}
     */
    saveGroup(groupInfo) {
        /** @type {?} */
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        return this.http.post(this.groupApiUrl, groupInfo, options);
    }
}
ApiService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ApiService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /** @type {?} */
    ApiService.prototype.absUrl;
    /** @type {?} */
    ApiService.prototype.groupApiUrl;
    /** @type {?} */
    ApiService.prototype.getOrgsUrl;
    /**
     * @type {?}
     * @private
     */
    ApiService.prototype.http;
}
