// @farris/component-querycondition/esm5/lib/directives/clear-fields.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/clear-fields.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Optional, Self, Injector } from '@angular/core';
import { NgControl } from '@angular/forms';
import { QueryConditionComponent } from '../querycondition.component';
var ClearFieldsDirective = /** @class */ (function () {
    function ClearFieldsDirective(ngCtrl, qcc, inject) {
        this.ngCtrl = ngCtrl;
        this.qcc = qcc;
        this.inject = inject;
        this.clearFields = '';
    }
    /**
     * @return {?}
     */
    ClearFieldsDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        //弹出帮助组件指定调用了父附件QueryConditionComponent，单独使用时无父组件
        if (!this.qcc)
            return;
        if (this.clearFields && this.ngCtrl) {
            this.ngCtrl.valueChanges.subscribe((/**
             * @param {?} v
             * @return {?}
             */
            function (v) {
                if (_this.qcc.canClearRelatedFields) {
                    _this.qcc.fieldCodeOfClearFields = [];
                    _this.clearFields.split(',').forEach((/**
                     * @param {?} n
                     * @return {?}
                     */
                    function (n) {
                        /** @type {?} */
                        var visibleConditions = _this.qcc.selectedFieldConditions.filter((/**
                         * @param {?} condition
                         * @return {?}
                         */
                        function (condition) { return !!condition.visible; }));
                        /** @type {?} */
                        var conditionItem = visibleConditions.find((/**
                         * @param {?} c
                         * @return {?}
                         */
                        function (c) { return c.fieldCode === n; }));
                        if (conditionItem) {
                            _this.qcc.fieldCodeOfClearFields.push(n);
                            conditionItem.value.clearValue();
                        }
                    }));
                }
            }));
        }
    };
    ClearFieldsDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[clear-fields]'
                },] }
    ];
    /** @nocollapse */
    ClearFieldsDirective.ctorParameters = function () { return [
        { type: NgControl, decorators: [{ type: Optional }, { type: Self }] },
        { type: QueryConditionComponent, decorators: [{ type: Optional }] },
        { type: Injector }
    ]; };
    ClearFieldsDirective.propDecorators = {
        clearFields: [{ type: Input, args: ['clear-fields',] }]
    };
    return ClearFieldsDirective;
}());
export { ClearFieldsDirective };
if (false) {
    /** @type {?} */
    ClearFieldsDirective.prototype.clearFields;
    /**
     * @type {?}
     * @private
     */
    ClearFieldsDirective.prototype.ngCtrl;
    /**
     * @type {?}
     * @private
     */
    ClearFieldsDirective.prototype.qcc;
    /**
     * @type {?}
     * @private
     */
    ClearFieldsDirective.prototype.inject;
}
