// @farris/component-querycondition/esm5/lib/directives/label-text-multi.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/label-text-multi.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Injector, ElementRef, NgZone } from '@angular/core';
var LabelTextMultiDirective = /** @class */ (function () {
    function LabelTextMultiDirective(el, injector) {
        this.el = el;
        this.injector = injector;
        this.multiCls = 'farris-group-multi-label';
        this._enableMulti = false;
        // 获取ngZone
        this.ngZone = null;
        if (!this.ngZone) {
            this.ngZone = this.injector.get(NgZone, null);
        }
    }
    Object.defineProperty(LabelTextMultiDirective.prototype, "enableMulti", {
        get: /**
         * @return {?}
         */
        function () {
            return this._enableMulti;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._enableMulti = value;
            this.judgeMulti();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    LabelTextMultiDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    LabelTextMultiDirective.prototype.judgeMulti = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this._enableMulti) {
            this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () {
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    /** @type {?} */
                    var textEl = _this.el.nativeElement.querySelector('.farris-label-text');
                    if (!textEl) {
                        return;
                    }
                    // 超出
                    if (textEl.clientWidth < textEl.scrollWidth) {
                        _this.el.nativeElement.className = _this.el.nativeElement.className + ' ' + _this.multiCls;
                    }
                    else {
                        _this.el.nativeElement.className = _this.el.nativeElement.className.replace(_this.multiCls, '');
                    }
                }), 20);
            }));
        }
    };
    LabelTextMultiDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[label-text-multi]'
                },] }
    ];
    /** @nocollapse */
    LabelTextMultiDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Injector }
    ]; };
    LabelTextMultiDirective.propDecorators = {
        enableMulti: [{ type: Input, args: ['label-text-multi',] }]
    };
    return LabelTextMultiDirective;
}());
export { LabelTextMultiDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    LabelTextMultiDirective.prototype.multiCls;
    /**
     * @type {?}
     * @private
     */
    LabelTextMultiDirective.prototype._enableMulti;
    /**
     * @type {?}
     * @private
     */
    LabelTextMultiDirective.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    LabelTextMultiDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    LabelTextMultiDirective.prototype.injector;
}
