// @gsp-svc/cloudprint/esm5/lib/entitys/printEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter } from '@angular/core';
var PrintEntity = /** @class */ (function () {
    function PrintEntity() {
        this.downLoadFileName = 'PrintReport';
        /**
         * 保存前事件
         */
        this.onSaveReportEvent = new EventEmitter();
        /**
         * 退出前事件
         * 可用于销毁资源
         */
        this.onExitEvent = new EventEmitter();
        /**
         * 保存成功之后的id事件
         */
        this.onGetPrintFormatIdEvent = new EventEmitter();
        /**
         * 保存成功后的事件
         */
        this.onSavedReportEvent = new EventEmitter();
    }
    return PrintEntity;
}());
export { PrintEntity };
if (false) {
    /**
     * 是否全屏渲染
     * 默认：false
     * @type {?}
     */
    PrintEntity.prototype.isFullScreen;
    /**
     * SessionID
     * @type {?}
     */
    PrintEntity.prototype.sessionId;
    /**
     * 第三方应用的基地址
     * (默认GSP云打印基地址)
     * @type {?}
     */
    PrintEntity.prototype.appUrl;
    /**
     * 当前语言
     * @type {?}
     */
    PrintEntity.prototype.language;
    /**
     * 一个开关，是否设计器处理保存逻辑
     * 默认为:false
     * @type {?}
     */
    PrintEntity.prototype.isUseMetaData;
    /**
     * 设计器的高度
     * @type {?}
     */
    PrintEntity.prototype.designerHeight;
    /**
     * 格式id
     * @type {?}
     */
    PrintEntity.prototype.foramtId;
    /**
     * 报表格式(设计时)
     * @type {?}
     */
    PrintEntity.prototype.format;
    /**
     * 数据Id
     * @type {?}
     */
    PrintEntity.prototype.dataId;
    /**
     * 报表数据
     * @type {?}
     */
    PrintEntity.prototype.data;
    /**
     * schema的类型：
     *  BE
     *  vo
     *  BO
     *  co
     * @type {?}
     */
    PrintEntity.prototype.schemaType;
    /** @type {?} */
    PrintEntity.prototype.printFormat;
    /** @type {?} */
    PrintEntity.prototype.downLoadFileName;
    /** @type {?} */
    PrintEntity.prototype.state;
    /**
     * 已经渲染成形的报表格式(运行时)
     * @type {?}
     */
    PrintEntity.prototype.rendedFormat;
    /**
     * 保存前事件
     * @type {?}
     */
    PrintEntity.prototype.onSaveReportEvent;
    /**
     * 退出前事件
     * 可用于销毁资源
     * @type {?}
     */
    PrintEntity.prototype.onExitEvent;
    /**
     * 保存成功之后的id事件
     * @type {?}
     */
    PrintEntity.prototype.onGetPrintFormatIdEvent;
    /**
     * 保存成功后的事件
     * @type {?}
     */
    PrintEntity.prototype.onSavedReportEvent;
}
