// @farris/ui-datatable/esm5/lib/datatable-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
var DTCheckboxComponent = /** @class */ (function () {
    function DTCheckboxComponent() {
        this.checkedChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    DTCheckboxComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} event
     * @return {?}
     */
    DTCheckboxComponent.prototype.handleClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this.disabled) {
            this.checkedChange.emit({ originalEvent: event, checked: !this.checked });
        }
    };
    DTCheckboxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dt-checkbox',
                    template: "\n        <div class=\"custom-control custom-checkbox  custom-control-inline dt-checkbox f-checkradio-single\">\n            <input type=\"checkbox\" class=\"custom-control-input\"  [checked]=\"checked\">\n            <label class=\"custom-control-label\" (click)=\"handleClick($event)\"></label>\n        </div>\n    "
                }] }
    ];
    /** @nocollapse */
    DTCheckboxComponent.ctorParameters = function () { return []; };
    DTCheckboxComponent.propDecorators = {
        checked: [{ type: Input }],
        id: [{ type: Input }],
        disabled: [{ type: Input }],
        checkedChange: [{ type: Output }]
    };
    return DTCheckboxComponent;
}());
export { DTCheckboxComponent };
if (false) {
    /** @type {?} */
    DTCheckboxComponent.prototype.checked;
    /** @type {?} */
    DTCheckboxComponent.prototype.id;
    /** @type {?} */
    DTCheckboxComponent.prototype.disabled;
    /** @type {?} */
    DTCheckboxComponent.prototype.checkedChange;
}
