// @farris/command-services/esm2015/lib/feature-edit/feature-edit.service.js
import { Injectable, Injector } from "@angular/core";
import { FeatureEditorService } from '@farris/ui-feature-editor';
import { FormNotifyService } from "../form-notify.service";
import { map, switchMap, tap } from "rxjs/operators";
import { FormLoadingService } from "../form-loading/form-loading.service";
import { FeatureDataService } from './feature-data.service';
import { LanguageService } from '../languag.service';
import { of } from "rxjs";
export class FeatureEditService {
    constructor(injector, featureEditorService, featureDataService, loadingService, notifyService, languageService) {
        this.injector = injector;
        this.featureEditorService = featureEditorService;
        this.featureDataService = featureDataService;
        this.loadingService = loadingService;
        this.notifyService = notifyService;
        this.languageService = languageService;
    }
    edit(materialId, materialFeatureId, options) {
        if (!materialId) {
            throw new Error('[FeatureEditService]物料id不能为空！');
        }
        if (!options) {
            options = {};
        }
        if (typeof options === 'string' && options.startsWith('{') && options.endsWith('}')) {
            options = JSON.parse(options);
        }
        options.getHelpInfo = this.featureDataService.getHelpInfo;
        this.loadingService.show();
        return this.featureDataService.getFeaturesByMaterialId(materialId).pipe(tap((result) => {
            this.loadingService.hide();
            const returnValue = result || null;
            if (returnValue) {
                const featureInfo = JSON.parse(returnValue);
                const props = featureInfo.props || null;
                if (!props || props.length < 1) {
                    this.notifyService.warning(this.languageService.propsIsEmpty);
                    return;
                }
                this.featureEditorService.show(props, options);
            }
            else {
                this.notifyService.error(this.languageService.propsIsEmpty);
            }
        }));
    }
    buildFeatures(materialId, materialFeatureId) {
        const features$ = this.featureDataService.getFeaturesByMaterialId(materialId);
        return features$.pipe(switchMap(featureSet => {
            if (materialFeatureId) {
                const configedFeatures$ = this.featureDataService.getConfigedValueByFeatureId(materialFeatureId, materialFeatureId);
                return configedFeatures$.pipe(map(defaultConfigs => {
                    const featureTemplate = featureSet.props;
                    const features = this.mergeFeatures(featureTemplate, defaultConfigs);
                    return features;
                }));
            }
            else {
                return of(featureSet.props);
            }
        }));
    }
    mergeFeatures(featureTemplate, defaultConfigs) {
        return null;
    }
}
FeatureEditService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FeatureEditService.ctorParameters = () => [
    { type: Injector },
    { type: FeatureEditorService },
    { type: FeatureDataService },
    { type: FormLoadingService },
    { type: FormNotifyService },
    { type: LanguageService }
];
