// @farris/dynamic-control-group/esm2015/lib/components/help/help.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { trim } from 'lodash-es';
import { CommonUtils } from '@farris/ui-common';
export class HelpComponent {
    /**
     * @param {?} utils
     */
    constructor(utils) {
        this.utils = utils;
        /**
         * @param helpData 选中的数据，如果是多选，是数组；如果是单选，是普通对象
         * @param mapFields
         * @param bindingData 帮助所在form的value对象
         */
        this.mapping = (/**
         * @param {?} helpData
         * @param {?} mapFields
         * @return {?}
         */
        (helpData, mapFields) => {
            console.log('helpData ', helpData);
            if (mapFields) {
                /** @type {?} */
                const helpFields = Object.keys(mapFields);
                helpFields.forEach((/**
                 * @param {?} f
                 * @return {?}
                 */
                (f) => {
                    /** @type {?} */
                    let val = '';
                    if (helpData) {
                        if (helpData instanceof Array) {
                            val = helpData.map((/**
                             * @param {?} h
                             * @return {?}
                             */
                            (h) => {
                                return this.utils.getValue(f, h);
                            })).join(',');
                        }
                        else {
                            val = this.utils.getValue(f, helpData);
                        }
                    }
                    mapFields[f].split(',').forEach((/**
                     * @param {?} ff
                     * @return {?}
                     */
                    (ff) => {
                        /** @type {?} */
                        const field = trim(ff);
                        // this.utils.setValue(bindingData, field, val);
                        this.form.controls[field].setValue(val);
                    }));
                }));
            }
        });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.options = this.control.control;
        this.className = this.control.className ? this.control.className : 'col-12 col-md-6 col-xl-3 col-el-2';
    }
}
HelpComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-help",
                template: `
    <div class="farris-group-wrap" [formGroup]="form">
      <div class="form-group farris-form-group">
        <label class="col-form-label">
          <span *ngIf="0" class="farris-label-info text-danger">*</span>
          <span class="farris-label-text">{{control.name}}</span>
        </label>
        <div class="farris-input-wrap">
            <farris-lookup-grid
              [formControlName]="control.labelCode"
              [uri]="options.uri"
              [singleSelect]="options.singleSelect"
              [displayType]="options.displayType"
              [idField]="options.idField"
              [textField]="options.textField"
              [valueField]="options.valueField"
              [title]="options.dialogTitle"
              [mappingFn]="mapping"
              [mapFields]="options.mapFields"
              [readonly]="options.readonly"
          ></farris-lookup-grid>
        </div>
      </div>
    </div>
  `
            }] }
];
/** @nocollapse */
HelpComponent.ctorParameters = () => [
    { type: CommonUtils }
];
HelpComponent.propDecorators = {
    className: [{ type: HostBinding, args: ['class',] }],
    control: [{ type: Input }],
    form: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    HelpComponent.prototype.className;
    /** @type {?} */
    HelpComponent.prototype.control;
    /** @type {?} */
    HelpComponent.prototype.form;
    /** @type {?} */
    HelpComponent.prototype.options;
    /**
     * \@param helpData 选中的数据，如果是多选，是数组；如果是单选，是普通对象
     * \@param mapFields
     * \@param bindingData 帮助所在form的value对象
     * @type {?}
     */
    HelpComponent.prototype.mapping;
    /**
     * @type {?}
     * @private
     */
    HelpComponent.prototype.utils;
}
