// @farris/dynamic-control-group/esm5/lib/directives/dynamic-control.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ComponentFactoryResolver, Directive, Injector, Input, ViewContainerRef } from "@angular/core";
import { DYNAMIC_CONTROL_TYPE } from "../interfaces/dynamic-control.interface";
import { InputComponent } from "../components/input/input.component";
import { FormGroup, FormControl } from '@angular/forms';
import { SingleDateComponent } from '../components/date/single-date.component';
import { DateRangeComponent } from '../components/date/date-range.component';
import { SingleNumberComponent } from '../components/number/single-number.component';
import { NumberRangeComponent } from '../components/number/number-range.component';
import { BaseDropdownComponent } from '../components/dropdown/base-dropdown.component';
import { HelpComponent } from '../components/help/help.component';
/** @type {?} */
var componentMapper = new Map();
componentMapper.set(DYNAMIC_CONTROL_TYPE.HELP, HelpComponent);
componentMapper.set(DYNAMIC_CONTROL_TYPE.BASE_DROPDOWN, BaseDropdownComponent);
componentMapper.set(DYNAMIC_CONTROL_TYPE.DATE_RANGE, DateRangeComponent);
componentMapper.set(DYNAMIC_CONTROL_TYPE.NUMBER_RANGE, NumberRangeComponent);
componentMapper.set(DYNAMIC_CONTROL_TYPE.SINGLE_DATE, SingleDateComponent);
componentMapper.set(DYNAMIC_CONTROL_TYPE.SINGLE_NUMBER, SingleNumberComponent);
componentMapper.set(DYNAMIC_CONTROL_TYPE.TEXT, InputComponent);
var DynamicFormControlDirective = /** @class */ (function () {
    function DynamicFormControlDirective(resolver, container, injector) {
        this.resolver = resolver;
        this.container = container;
        this.injector = injector;
    }
    /**
     * @return {?}
     */
    DynamicFormControlDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.formGroup = this.addToForm(this.data, this.formGroup);
        /** @type {?} */
        var factory = this.resolver.resolveComponentFactory(componentMapper.get(this.data.control.controlType));
        this.componentRef = this.container.createComponent(factory);
        this.componentRef.instance.form = this.formGroup;
        this.componentRef.instance.control = this.data;
        // const form = this.addToForm(this.data, this.formGroup);
    };
    /**
     * @param {?} control
     * @param {?} originalForm
     * @return {?}
     */
    DynamicFormControlDirective.prototype.addToForm = /**
     * @param {?} control
     * @param {?} originalForm
     * @return {?}
     */
    function (control, originalForm) {
        //originalForm 外部传入FormGroup的实例
        originalForm.addControl(control.labelCode, new FormControl(control.value));
        if (control.control.controlType === DYNAMIC_CONTROL_TYPE.HELP) {
            /** @type {?} */
            var helpControl = (/** @type {?} */ (control));
            originalForm.addControl(helpControl.control.hiddenInputName, new FormControl(''));
        }
        return originalForm;
    };
    DynamicFormControlDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[dynamicFormControl]"
                },] }
    ];
    /** @nocollapse */
    DynamicFormControlDirective.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: ViewContainerRef },
        { type: Injector }
    ]; };
    DynamicFormControlDirective.propDecorators = {
        data: [{ type: Input }],
        formGroup: [{ type: Input }]
    };
    return DynamicFormControlDirective;
}());
export { DynamicFormControlDirective };
if (false) {
    /** @type {?} */
    DynamicFormControlDirective.prototype.data;
    /** @type {?} */
    DynamicFormControlDirective.prototype.formGroup;
    /** @type {?} */
    DynamicFormControlDirective.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    DynamicFormControlDirective.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    DynamicFormControlDirective.prototype.container;
    /**
     * @type {?}
     * @private
     */
    DynamicFormControlDirective.prototype.injector;
}
