// @gsp-cmp/coderule-selector/esm2015/lib/service/code-rule-select.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/code-rule-select.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpHeaders, HttpClient } from '@angular/common/http';
export class CodeRuleSelectService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.QUERY_URL = "/api/runtime/bcc/v1.0/code-rule-data";
        this.metadatas = [];
    }
    /**
     * @param {?} metadata
     * @return {?}
     */
    addMetadataRT(metadata) {
        this.metadatas.push(metadata);
    }
    /**
     * @return {?}
     */
    clearData() {
        this.metadatas = [];
        this.codeRule = null;
    }
    // 根据类型获取Cloud Server下所有的元数据
    /**
     * @return {?}
     */
    getCodeRuleList() {
        /** @type {?} */
        let headers = new HttpHeaders();
        headers.append("sessionId", this.sessionId);
        return this.http.get(`${this.QUERY_URL}?serviceUnit=${this.serviceUnit}`, { headers: headers });
    }
}
CodeRuleSelectService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CodeRuleSelectService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /** @type {?} */
    CodeRuleSelectService.prototype.serviceUnit;
    /** @type {?} */
    CodeRuleSelectService.prototype.sessionId;
    /**
     * @type {?}
     * @private
     */
    CodeRuleSelectService.prototype.QUERY_URL;
    /** @type {?} */
    CodeRuleSelectService.prototype.codeRule;
    /** @type {?} */
    CodeRuleSelectService.prototype.metadatas;
    /**
     * @type {?}
     * @private
     */
    CodeRuleSelectService.prototype.http;
}
