// @farris/ui-dropdown/esm2015/lib/component/dropdown.item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
export class FarrisDropdownItemComponent {
    constructor() {
        this.select = new EventEmitter();
    }
    /**
     * @return {?}
     */
    selectItem() {
        if (this.model.disabled) {
            return;
        }
        this.select.emit(this.model);
    }
}
FarrisDropdownItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'farris-dropdown-item',
                template: `
        <div class="dropdown-divider" *ngIf="model.divide"></div>
        <a
            class="dropdown-item"
            [class.active]="model.active"
            [class.disabled]="model.disabled"
            href="javascript:void(0)"
            (click)="selectItem()"
            >{{ model.label }}</a
        >
    `
            }] }
];
FarrisDropdownItemComponent.propDecorators = {
    model: [{ type: Input }],
    select: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FarrisDropdownItemComponent.prototype.model;
    /** @type {?} */
    FarrisDropdownItemComponent.prototype.select;
}
