// @edp-pmf/grapheditor/dist/cjs/lib/util/ResourceKeys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceKeys = void 0;
/**
 * 资源Key列表
 */
var ResourceKeys;
(function (ResourceKeys) {
    //temp = "temp",
    // File
    ResourceKeys["pmf_file"] = "file";
    ResourceKeys["pmf_save"] = "save";
    ResourceKeys["pmf_saveAs"] = "saveAs";
    ResourceKeys["pmf_export"] = "export";
    //Edit
    ResourceKeys["pmf_edit"] = "edit";
    ResourceKeys["pmf_undo"] = "undo";
    ResourceKeys["pmf_redo"] = "redo";
    ResourceKeys["pmf_cut"] = "cut";
    ResourceKeys["pmf_copy"] = "copy";
    ResourceKeys["pmf_paste"] = "paste";
    ResourceKeys["pmf_pasteHere"] = "pasteHere";
    ResourceKeys["pmf_delete"] = "delete";
    ResourceKeys["pmf_deleteAll"] = "deleteAll";
    ResourceKeys["pmf_duplicate"] = "duplicate";
    ResourceKeys["pmf_selectVertices"] = "selectVertices";
    ResourceKeys["pmf_selectEdges"] = "selectEdges";
    ResourceKeys["pmf_selectAll"] = "selectAll";
    ResourceKeys["pmf_selectNone"] = "selectNone";
    ResourceKeys["pmf_lockUnlock"] = "lockUnlock";
    //View
    ResourceKeys["pmf_view"] = "view";
    ResourceKeys["pmf_formatPanel"] = "formatPanel";
    ResourceKeys["pmf_outline"] = "outline";
    ResourceKeys["pmf_layers"] = "layers";
    ResourceKeys["pmf_pageView"] = "pageView";
    ResourceKeys["pmf_grid"] = "grid";
    ResourceKeys["pmf_connectionArrows"] = "connectionArrows";
    ResourceKeys["pmf_connectionPoints"] = "connectionPoints";
    ResourceKeys["pmf_resetView"] = "resetView";
    ResourceKeys["pmf_zoom"] = "zoom";
    ResourceKeys["pmf_zoomIn"] = "zoomIn";
    ResourceKeys["pmf_zoomOut"] = "zoomOut";
    ResourceKeys["pmf_fitWindow"] = "fitWindow";
    ResourceKeys["pmf_fitPage"] = "fitPage";
    ResourceKeys["pmf_fitTwoPages"] = "fitTwoPages";
    ResourceKeys["pmf_fitPageWidth"] = "fitPageWidth";
    ResourceKeys["pmf_customZoom"] = "custom";
    // Arrange
    ResourceKeys["pmf_arrange"] = "arrange";
    ResourceKeys["pmf_toFront"] = "toFront";
    ResourceKeys["pmf_toBack"] = "toBack";
    ResourceKeys["pmf_align"] = "align";
    ResourceKeys["pmf_leftAlign"] = "leftAlign";
    ResourceKeys["pmf_center"] = "center";
    ResourceKeys["pmf_rightAlign"] = "rightAlign";
    ResourceKeys["pmf_topAlign"] = "topAlign";
    ResourceKeys["pmf_middle"] = "middle";
    ResourceKeys["pmf_bottomAlign"] = "bottomAlign";
    ResourceKeys["pmf_distribute"] = "distribute";
    ResourceKeys["pmf_horizontal"] = "horizontal";
    ResourceKeys["pmf_vertical"] = "vertical";
    ResourceKeys["pmf_clearWaypoints"] = "clearWaypoints";
    ResourceKeys["pmf_autosize"] = "autosize";
    //Help
    ResourceKeys["pmf_help"] = "help";
    ResourceKeys["pmf_about"] = "about";
    //pmf------------------
    ResourceKeys["pmf_mainMenu"] = "pmf_mainMenu";
    ResourceKeys["pmf_viewModelXml"] = "pmf_viewModelXml";
})(ResourceKeys = exports.ResourceKeys || (exports.ResourceKeys = {}));
