// @farris/oa-command-services/esm2015/lib/task-center-bridge/entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/** @enum {string} */
const WFActionType = {
    /**
     * 审批通过
    */
    Pass: 'Pass',
    /**
     *
    */
    Resubmit: 'Resubmit',
    /**
     * 驳回
    */
    Back: 'Back',
    /**
     * 移交
    */
    Transfer: 'Transfer',
    /**
     * 评论
    */
    Comment: 'Comment',
};
export { WFActionType };
/**
 * @record
 */
export function Parameter() { }
if (false) {
    /** @type {?} */
    Parameter.prototype.code;
    /** @type {?} */
    Parameter.prototype.value;
}
/**
 * @record
 */
export function Action() { }
if (false) {
    /** @type {?} */
    Action.prototype.code;
    /** @type {?} */
    Action.prototype.parameters;
}
/**
 * @record
 */
export function WFTask() { }
if (false) {
    /** @type {?} */
    WFTask.prototype.id;
    /** @type {?} */
    WFTask.prototype.sourceId;
    /** @type {?} */
    WFTask.prototype.name;
    /** @type {?} */
    WFTask.prototype.state;
    /** @type {?} */
    WFTask.prototype.owner;
    /** @type {?} */
    WFTask.prototype.priority;
    /** @type {?} */
    WFTask.prototype.categoryId;
    /** @type {?} */
    WFTask.prototype.submitter;
    /** @type {?} */
    WFTask.prototype.startTime;
    /** @type {?} */
    WFTask.prototype.endTime;
    /** @type {?} */
    WFTask.prototype.processInstanceId;
    /** @type {?} */
    WFTask.prototype.refBizId;
    /** @type {?} */
    WFTask.prototype.submitterName;
    /** @type {?} */
    WFTask.prototype.bizCategory;
    /** @type {?} */
    WFTask.prototype.activityType;
    /** @type {?} */
    WFTask.prototype.activityDefinitionId;
    /** @type {?} */
    WFTask.prototype.activityInstanceId;
    /** @type {?} */
    WFTask.prototype.fromType;
    /** @type {?} */
    WFTask.prototype.tenantId;
    /** @type {?} */
    WFTask.prototype.readTime;
    /** @type {?} */
    WFTask.prototype.createdTime;
    /** @type {?} */
    WFTask.prototype.creator;
    /** @type {?} */
    WFTask.prototype.lastModifiedTime;
    /** @type {?} */
    WFTask.prototype.lastModifier;
    /** @type {?} */
    WFTask.prototype.actions;
}
