// @gsp-wf/wf-process-editor/esm5/lib/views/performer/views/performer-edit/performer-edit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, HostBinding, ViewChild, Input, Optional, LOCALE_ID, Inject } from '@angular/core';
import { UserPerformerComponent } from '../user-performer/user-performer.component';
import { PositionPerformerComponent } from '../position-performer/position-performer.component';
import { ComponentPerformerComponent } from '../component-performer/component-performer.component';
import { ReportingLineComponent } from '../reporting-line/reporting-line.component';
import { MatrixPerformerComponent } from '../matrix-performer/matrix-performer.component';
import { HttpService } from '@ecp-caf/caf-common';
import { ActivityExecutorPerformerComponent } from '../activity-executor-performer/activity-executor-performer.component';
import { ProcessDesignerUIState } from '../../../../app/process-designer.uistate';
import { ExprEditMode, ExpressionService, InjectedType } from '@gsp-svc/expression';
import { OrganizationParticipantExpression, ActionScope, OrganizationLevel } from '../../../../domain/entities/organization-participant-expression';
import { VariablePerformerComponent } from '../variable-performer/variable-performer.component';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../../../i18n/pipe/wf-process-editor-locale.pipe';
var PerformerEditComponent = /** @class */ (function () {
    function PerformerEditComponent(designerUIState, expr, http, localeId, resources, localePipe) {
        this.designerUIState = designerUIState;
        this.expr = expr;
        this.http = http;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.cls = 'farris-main-area flex-column';
        this.tab = 'user';
        this.display = true;
        this.organizationLevelCollection = [
            {
                value: 'UNIT',
                text: this.localePipe.transform('component.participant.currentUnit'),
                orgId: 'currentUnit'
            }, {
                value: 'DEPARTMENT',
                text: this.localePipe.transform('component.participant.currentDepartment'),
                orgId: 'currentDepartment'
            }, {
                value: 'VARORGANIZATION',
                text: this.localePipe.transform('component.participant.variableOrganization'),
                orgId: ''
            }
        ];
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
        this.organizationParticipant = new OrganizationParticipantExpression();
        this.organizationParticipant.actionScope = ActionScope.AssignUser;
        this.organizationParticipant.organizationLevel = OrganizationLevel.DEPARTMENT;
        this.organizationParticipant.organizationId = 'currentDepartment';
        this.organizationParticipant.organizationName = this.localePipe.transform('component.participant.currentDepartment');
        this.organizationParticipant.includeChildren = false;
        this.organizationParticipant.tag = 'variable';
    }
    /**
     * @return {?}
     */
    PerformerEditComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    PerformerEditComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.performer) {
            switch (this.performer.type) {
                case 'wf.UserParticipantExpression':
                    this.usersCmp.users = [];
                    this.usersCmp.users.push(this.performer.value);
                    // 2023-5-23 用户参与者支持编辑时单选+显示已选用户；
                    this.usersCmp.multiSelect = this.editState === 'edit' ? false : true;
                    // 考虑到兼容历史数据，暂时不使用用户参与者实体类中的userType属性
                    if (this.performer.value.performerId && this.performer.value.performerId.charAt(0) === '{') {
                        /** @type {?} */
                        var url = '/api/runtime/wf/v1.0/reportingOrgType/hasAdminOrg';
                        this.http.get(url).subscribe((/**
                         * @param {?} data
                         * @return {?}
                         */
                        function (data) {
                            _this.usersCmp.userTypeButtonNeedShow = data;
                            _this.usersCmp.userType = 'variable';
                            _this.usersCmp.curOrgType = _this.performer.value.orgType === 'Sys' ? 'Sys' : 'Administrative_organization';
                            _this.usersCmp.condition = _this.performer.value.condition;
                            // 根据表达式中的函数名判断是否使用的是上下文变量
                            /** @type {?} */
                            var isContextVariable = _this.performer.value.performerId.indexOf('GetContextParameter') >= 0;
                            _this.usersCmp.variableOrgTypes = isContextVariable ? [
                                { name: '系统用户', value: 'Sys' }
                            ] : [
                                { name: '系统用户', value: 'Sys' },
                                { name: '员工', value: 'Administrative_organization' }
                            ];
                        }));
                    }
                    else {
                        this.usersCmp.userType = 'user';
                        this.usersCmp.condition = this.performer.value.condition;
                    }
                    break;
                case 'wf.VariableParticipantExpression':
                    // TODO:参与者只保留一份最好
                    this.varCmp.users = [];
                    this.varCmp.users.push(this.performer.value);
                    this.varCmp.user = this.performer.value;
                    this.varCmp.editState = this.editState === 'edit' ? false : true;
                    // 判断是变量参与者
                    if (this.performer.value.varValue && this.performer.value.varValue.charAt(0) === '{') {
                        // 根据表达式中的函数名判断是否使用的是上下文变量
                        /** @type {?} */
                        var isContextVariable = this.performer.value.varValue.indexOf('GetContextParameter') >= 0;
                        // 上下文变量的变量值类型只有系统用户id
                        this.varCmp.variableOrgTypes = isContextVariable ? [
                            { name: '用户ID', value: 'Sys' }
                        ] : [
                            { name: '用户ID', value: 'Sys' },
                            { name: '员工ID', value: 'Administrative_staff' },
                        ];
                        // 如果是insuite场景还要增加行政组织id
                        this.varCmp.curOrgType = this.performer.value.varType;
                        this.varCmp.relationShipType = this.performer.value.relation;
                        this.varCmp.condition = this.performer.value.condition;
                    }
                    break;
                case 'wf.PositionParticipantExpression':
                    this.posiCmp.editState = this.editState === 'edit' ? false : true;
                    this.posiCmp.position = Object.assign({}, this.performer.value);
                    this.posiCmp.performerName = this.posiCmp.position.performerName;
                    if (this.posiCmp.position.positionType === '0') {
                        this.posiCmp.newFilter = { positionType: '0' };
                    }
                    else {
                        this.posiCmp.newFilter = { positionType: '1' };
                    }
                    break;
                case 'wf.ReportingLineParticipantExpression':
                    // this.tab = 'report';
                    this.reportingLineCmp.setReportingLineDataEntity = this.performer.value;
                    break;
                case 'wf.ComponentParticipantExpression':
                    // this.tab = 'component';
                    this.cmpCmp.component = this.performer.value;
                    break;
                case 'wf.MatrixParticipantExpression':
                    // this.tab = 'matrix';
                    this.matrixCmp.setMatrixDataEntity = this.performer.value;
                    break;
                case 'wf.ActivityExecutorParticipantExpression':
                    // this.tab = 'activityExecutor';
                    this.activityExecutorCmp.activityExecutors = [];
                    this.activityExecutorCmp.activityExecutors.push(this.performer.value);
                    this.activityExecutorCmp.condition = this.performer.value.condition;
                    break;
                case 'wf.OrganizationParticipantExpression':
                    if (this.performer.value.organizationLevel === OrganizationLevel.DEFAULT && this.performer.value.tag === 'expression') {
                        this.performer.value.organizationLevel = OrganizationLevel.VARORGANIZATION;
                    }
                    this.organizationParticipant = this.performer.value;
                    if (this.performer.value.OrganizationLevel === OrganizationLevel.DEPARTMENT && this.performer.value.tag === 'expression') {
                        this.performer.value.OrganizationLevel === OrganizationLevel.VARORGANIZATION;
                    }
                    break;
            }
        }
        if (this.performerType) {
            switch (this.performerType) {
                case 'wf.UserParticipantExpression':
                    this.usersCmp.userType = 'user';
                    break;
            }
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PerformerEditComponent.prototype.valueChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        switch (event.value) {
            case this.organizationLevelCollection[0].value:
                this.organizationParticipant.organizationId = this.organizationLevelCollection[0].orgId;
                this.organizationParticipant.organizationName = this.organizationLevelCollection[0].text;
                this.organizationParticipant.includeChildren = true;
                break;
            case this.organizationLevelCollection[1].value:
                this.organizationParticipant.organizationId = this.organizationLevelCollection[1].orgId;
                this.organizationParticipant.organizationName = this.organizationLevelCollection[1].text;
                break;
            case this.organizationLevelCollection[2].value:
                this.organizationParticipant.organizationId = this.organizationLevelCollection[2].orgId;
                this.organizationParticipant.organizationName = this.organizationLevelCollection[2].text;
                this.organizationParticipant.includeChildren = true;
                this.organizationParticipant.tag = 'expression';
                break;
        }
    };
    /**
     * @return {?}
     */
    PerformerEditComponent.prototype.getPerformers = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var users = this.usersCmp ? this.usersCmp.getPerformers() : [];
        /** @type {?} */
        var variable = this.varCmp ? this.varCmp.getPerformers() : null;
        /** @type {?} */
        var positions = this.posiCmp ? this.posiCmp.getPerformers() : [];
        /** @type {?} */
        var component = this.cmpCmp ? this.cmpCmp.getPerformers() : null;
        /** @type {?} */
        var reportingLine = this.reportingLineCmp ? this.reportingLineCmp.getPerformers() : null;
        /** @type {?} */
        var matrix = this.matrixCmp ? this.matrixCmp.getPerformers() : null;
        /** @type {?} */
        var activityExecutors = this.activityExecutorCmp ? this.activityExecutorCmp.getPerformers() : [];
        /** @type {?} */
        var userPerformers = [];
        // 用户参与者
        /** @type {?} */
        var activityExecutorPerformers = [];
        if ((users && users.length > 0)) {
            userPerformers = userPerformers.concat(users);
        }
        if (activityExecutors && activityExecutors.length > 0) {
            activityExecutorPerformers = activityExecutorPerformers.concat(activityExecutors);
        }
        /** @type {?} */
        var organizationParticipant;
        if (this.performerType === 'wf.OrganizationParticipantExpression') {
            if (this.organizationParticipant.organizationLevel === OrganizationLevel.VARORGANIZATION && this.organizationParticipant.tag === 'expression') {
                this.organizationParticipant.organizationLevel = OrganizationLevel.DEFAULT;
            }
            organizationParticipant = this.organizationParticipant;
        }
        return { userPerformers: userPerformers, variable: variable, positions: positions, component: component, reportingLine: reportingLine, matrix: matrix, activityExecutorPerformers: activityExecutorPerformers, organizationParticipant: organizationParticipant };
    };
    /**
     * @return {?}
     */
    PerformerEditComponent.prototype.editConditon = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var e_1, _a, e_2, _b, e_3, _c, e_4, _d;
        // this.expr.clearContext();
        if (this.designerUIState.schemas.length > 0) {
            try {
                for (var _e = tslib_1.__values(this.designerUIState.schemas), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var schema = _f.value;
                    this.expr.addSchema('Schema', schema);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_a = _e.return)) _a.call(_e);
                }
                finally { if (e_1) throw e_1.error; }
            }
            if (this.designerUIState.schemas[0].entityTypes && this.designerUIState.schemas[0].entityTypes.length > 0) {
                /** @type {?} */
                var entityType = this.designerUIState.schemas[0].entityTypes[0];
                this.expr.addInjectedEntity(entityType.name, 'Schema', entityType.name, InjectedType.EntityType);
            }
        }
        /** @type {?} */
        var context = new Array();
        context = this.designerUIState.addContext(context);
        if (this.designerUIState.subActivityContext.length > 0) {
            var _loop_1 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _g = tslib_1.__values(this.designerUIState.subActivityContext), _h = _g.next(); !_h.done; _h = _g.next()) {
                    var variable = _h.value;
                    _loop_1(variable);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_h && !_h.done && (_b = _g.return)) _b.call(_g);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        if (this.designerUIState.activityComponentContext.length > 0) {
            var _loop_2 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _j = tslib_1.__values(this.designerUIState.activityComponentContext), _k = _j.next(); !_k.done; _k = _j.next()) {
                    var variable = _k.value;
                    _loop_2(variable);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_k && !_k.done && (_c = _j.return)) _c.call(_j);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        if (context.length > 0) {
            try {
                for (var context_1 = tslib_1.__values(context), context_1_1 = context_1.next(); !context_1_1.done; context_1_1 = context_1.next()) {
                    var v = context_1_1.value;
                    this.expr.addContext(v);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (context_1_1 && !context_1_1.done && (_d = context_1.return)) _d.call(context_1);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
        this.expr.setEditPattern(ExprEditMode.Conditional); // 表达式展现模式：默认简易版
        this.expr.addExpressionText(this.organizationParticipant.condition);
        this.expr.buildExpression().then((/**
         * @param {?} expressioninfo
         * @return {?}
         */
        function (expressioninfo) {
            _this.organizationParticipant.condition = expressioninfo;
            //点击确定后，去除表达式里面的新增的上下文变量
            if (context.length > 0) {
                _this.expr.exprEntity.contextEntities = _this.expr.exprEntity.contextEntities.filter((/**
                 * @param {?} x
                 * @return {?}
                 */
                function (x) { return !context.some((/**
                 * @param {?} y
                 * @return {?}
                 */
                function (y) { return y.key === x.key; })); }));
            }
        }));
    };
    /**
     * @return {?}
     */
    PerformerEditComponent.prototype.editOrganizationId = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var e_5, _a, e_6, _b, e_7, _c, e_8, _d;
        // this.expr.clearContext();
        if (this.designerUIState.schemas.length > 0) {
            try {
                for (var _e = tslib_1.__values(this.designerUIState.schemas), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var schema = _f.value;
                    this.expr.addSchema('Schema', schema);
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_a = _e.return)) _a.call(_e);
                }
                finally { if (e_5) throw e_5.error; }
            }
            if (this.designerUIState.schemas[0].entityTypes && this.designerUIState.schemas[0].entityTypes.length > 0) {
                /** @type {?} */
                var entityType = this.designerUIState.schemas[0].entityTypes[0];
                this.expr.addInjectedEntity(entityType.name, 'Schema', entityType.name, InjectedType.EntityType);
            }
        }
        /** @type {?} */
        var context = new Array();
        context = this.designerUIState.addContext(context);
        if (this.designerUIState.subActivityContext.length > 0) {
            var _loop_3 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _g = tslib_1.__values(this.designerUIState.subActivityContext), _h = _g.next(); !_h.done; _h = _g.next()) {
                    var variable = _h.value;
                    _loop_3(variable);
                }
            }
            catch (e_6_1) { e_6 = { error: e_6_1 }; }
            finally {
                try {
                    if (_h && !_h.done && (_b = _g.return)) _b.call(_g);
                }
                finally { if (e_6) throw e_6.error; }
            }
        }
        if (this.designerUIState.activityComponentContext.length > 0) {
            var _loop_4 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _j = tslib_1.__values(this.designerUIState.activityComponentContext), _k = _j.next(); !_k.done; _k = _j.next()) {
                    var variable = _k.value;
                    _loop_4(variable);
                }
            }
            catch (e_7_1) { e_7 = { error: e_7_1 }; }
            finally {
                try {
                    if (_k && !_k.done && (_c = _j.return)) _c.call(_j);
                }
                finally { if (e_7) throw e_7.error; }
            }
        }
        if (context.length > 0) {
            try {
                for (var context_2 = tslib_1.__values(context), context_2_1 = context_2.next(); !context_2_1.done; context_2_1 = context_2.next()) {
                    var v = context_2_1.value;
                    this.expr.addContext(v);
                }
            }
            catch (e_8_1) { e_8 = { error: e_8_1 }; }
            finally {
                try {
                    if (context_2_1 && !context_2_1.done && (_d = context_2.return)) _d.call(context_2);
                }
                finally { if (e_8) throw e_8.error; }
            }
        }
        this.expr.setEditPattern(ExprEditMode.Advanced);
        this.expr.addExpressionText(this.organizationParticipant.organizationId);
        this.expr.buildExpression().then((/**
         * @param {?} expressioninfo
         * @return {?}
         */
        function (expressioninfo) {
            _this.organizationParticipant.organizationId = expressioninfo;
            //点击确定后，去除表达式里面的新增的上下文变量
            if (context.length > 0) {
                _this.expr.exprEntity.contextEntities = _this.expr.exprEntity.contextEntities.filter((/**
                 * @param {?} x
                 * @return {?}
                 */
                function (x) { return !context.some((/**
                 * @param {?} y
                 * @return {?}
                 */
                function (y) { return y.key === x.key; })); }));
            }
        }));
    };
    PerformerEditComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-performer-edit',
                    template: "<app-user-performer *ngIf=\"performerType === 'wf.UserParticipantExpression'\"></app-user-performer>\r\n<app-variable-performer *ngIf=\"performerType === 'wf.VariableParticipantExpression'\"></app-variable-performer>\r\n<app-position-performer *ngIf=\"performerType === 'wf.PositionParticipantExpression'\"></app-position-performer>\r\n<app-reporting-line *ngIf=\"performerType === 'wf.ReportingLineParticipantExpression'\"></app-reporting-line>\r\n<app-matrix-performer *ngIf=\"performerType === 'wf.MatrixParticipantExpression'\"></app-matrix-performer>\r\n<app-activity-handler-performer\r\n  *ngIf=\"performerType === 'wf.ActivityExecutorParticipantExpression'\"></app-activity-handler-performer>\r\n<app-component-performer [component]=\"component\" style=\"background: #FFF;\"\r\n  *ngIf=\"performerType === 'wf.ComponentParticipantExpression'\"></app-component-performer>\r\n\r\n<div *ngIf=\"performerType === 'wf.OrganizationParticipantExpression'\" class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\" style=\"background: #FFF;\">\r\n    <div class=\"f-utils-fill-flex-column farris-split-section mx-1 h-100\"\r\n      style=\"border-radius:16px;border: 1.5px solid #F1F3F7;\">\r\n      <form style=\"width: 100%; height: 100%;\">\r\n        <div class=\"d-flex m-2 flex-column w-100\">\r\n          <div class=\"d-flex flex-column\" style=\"height: 120px;\">\r\n            <div class=\"farris-form form-inline  farris-form-inline mr-5 my-2 pt-3\">\r\n              <div class=\"col-12 farris-group-auto\">\r\n                <div class=\"farris-group-wrap\">\r\n                  <div class=\"form-group farris-form-group\">\r\n                    <label class=\"col-form-label\">\r\n                      <span class=\"farris-label-text\">{{'component.participant.alternativeScope'|wfLocale}}</span>\r\n                    </label>\r\n                    <div class=\"farris-input-wrap ml-1\">\r\n                      <farris-combo-list [panelHeight]=\"100\" [(ngModel)]=\"organizationParticipant.organizationLevel\"\r\n                        name=\"actionScope\" [idField]=\"'value'\" [enableClear]=\"false\" [textField]=\"'text'\"\r\n                        [valueField]=\"'value'\" [data]=\"organizationLevelCollection\" [readonly]=\"false\"\r\n                        (valueChange)=\"valueChange($event)\">\r\n                      </farris-combo-list>\r\n                      <div class=\"farris-feedback valid-feedback\"></div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n              <!-- \u5355\u4F4D/\u90E8\u95E8\u53C2\u4E0E\u8005\u652F\u6301\u5355\u4F4DID -->\r\n              <div *ngIf=\"organizationParticipant.organizationLevel===organizationLevelCollection[2].value\"\r\n                class=\"col-12 farris-group-auto my-3 pr-3\">\r\n                <div class=\"farris-group-wrap\">\r\n                  <div class=\"form-group farris-form-group\">\r\n                    <label class=\"col-form-label\">\r\n                      <span class=\"farris-label-text\">{{'component.participant.orgId'|wfLocale}}</span>\r\n                    </label>\r\n                    <div class=\"farris-input-wrap ml-1\">\r\n                      <div class=\"input-group\">\r\n                        <input type=\"text\" class=\"form-control\" [(ngModel)]=\"organizationParticipant.organizationId\"\r\n                          name=\"organizationId\" [readOnly]=\"true\" />\r\n                        <div class=\"input-group-append\">\r\n                          <span class=\"input-group-text\" (click)=\"editOrganizationId()\" style=\"background: #FFF;\">\r\n                            <i class=\"f-icon f-icon-lookup\"></i>\r\n                          </span>\r\n                        </div>\r\n                      </div>\r\n                      <div class=\"farris-feedback valid-feedback\"></div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n              <div *ngIf=\"organizationParticipant.organizationLevel!=organizationLevelCollection[0].value\"\r\n                class=\"col-12 farris-group-auto my-3 pr-5\">\r\n                <div class=\"farris-group-wrap\">\r\n                  <div class=\"form-group farris-form-group\">\r\n                    <label class=\"col-form-label\">\r\n                      <span class=\"farris-label-text\">{{'component.participant.includeChildren'|wfLocale}}</span>\r\n                    </label>\r\n                    <div class=\"farris-input-wrap ml-1\">\r\n                      <farris-switch [(ngModel)]=\"organizationParticipant.includeChildren\"\r\n                        name=\"includeChildren\"></farris-switch>\r\n                      <div class=\"farris-feedback valid-feedback\"></div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"d-flex flex-column m-2 w-100\">\r\n            <div class=\"farris-form form-inline farris-form-inline mr-5\">\r\n              <div class=\"col-12 farris-group-auto my-4\" style=\"padding-left: 6px;padding-right: 21px;\">\r\n                <div class=\"farris-group-wrap\">\r\n                  <div class=\"form-group farris-form-group\">\r\n                    <label class=\"col-form-label\">\r\n                      <span class=\"farris-label-text\">{{'component.participant.conditon'|wfLocale}}</span>\r\n                    </label>\r\n                    <div class=\"farris-input-wrap\">\r\n                      <div class=\"input-group\">\r\n                        <input type=\"text\" class=\"form-control\" [(ngModel)]=\"organizationParticipant.condition\"\r\n                          name=\"conditon\" [readOnly]=\"true\" />\r\n                        <div class=\"input-group-append\">\r\n                          <span class=\"input-group-text\" (click)=\"editConditon()\" style=\"background: #FFF;\">\r\n                            <i class=\"f-icon f-icon-lookup\"></i>\r\n                          </span>\r\n                        </div>\r\n                      </div>\r\n                      <div class=\"farris-feedback valid-feedback\"></div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </form>\r\n    </div>\r\n  </div>\r\n</div>",
                    providers: [
                        WfLocalePipe,
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    PerformerEditComponent.ctorParameters = function () { return [
        { type: ProcessDesignerUIState },
        { type: ExpressionService },
        { type: HttpService },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
        { type: WfLocalePipe, decorators: [{ type: Optional }] }
    ]; };
    PerformerEditComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        display: [{ type: Input }],
        usersCmp: [{ type: ViewChild, args: [UserPerformerComponent,] }],
        varCmp: [{ type: ViewChild, args: [VariablePerformerComponent,] }],
        posiCmp: [{ type: ViewChild, args: [PositionPerformerComponent,] }],
        cmpCmp: [{ type: ViewChild, args: [ComponentPerformerComponent,] }],
        reportingLineCmp: [{ type: ViewChild, args: [ReportingLineComponent,] }],
        matrixCmp: [{ type: ViewChild, args: [MatrixPerformerComponent,] }],
        activityExecutorCmp: [{ type: ViewChild, args: [ActivityExecutorPerformerComponent,] }]
    };
    return PerformerEditComponent;
}());
export { PerformerEditComponent };
if (false) {
    /** @type {?} */
    PerformerEditComponent.prototype.cls;
    /** @type {?} */
    PerformerEditComponent.prototype.component;
    /** @type {?} */
    PerformerEditComponent.prototype.performer;
    /** @type {?} */
    PerformerEditComponent.prototype.performerType;
    /** @type {?} */
    PerformerEditComponent.prototype.userType;
    /** @type {?} */
    PerformerEditComponent.prototype.editState;
    /** @type {?} */
    PerformerEditComponent.prototype.tab;
    /** @type {?} */
    PerformerEditComponent.prototype.display;
    /** @type {?} */
    PerformerEditComponent.prototype.organizationLevelCollection;
    /** @type {?} */
    PerformerEditComponent.prototype.organizationParticipant;
    /** @type {?} */
    PerformerEditComponent.prototype.usersCmp;
    /** @type {?} */
    PerformerEditComponent.prototype.varCmp;
    /** @type {?} */
    PerformerEditComponent.prototype.posiCmp;
    /** @type {?} */
    PerformerEditComponent.prototype.cmpCmp;
    /** @type {?} */
    PerformerEditComponent.prototype.reportingLineCmp;
    /** @type {?} */
    PerformerEditComponent.prototype.matrixCmp;
    /** @type {?} */
    PerformerEditComponent.prototype.activityExecutorCmp;
    /**
     * @type {?}
     * @private
     */
    PerformerEditComponent.prototype.designerUIState;
    /**
     * @type {?}
     * @private
     */
    PerformerEditComponent.prototype.expr;
    /**
     * @type {?}
     * @private
     */
    PerformerEditComponent.prototype.http;
    /**
     * @type {?}
     * @private
     */
    PerformerEditComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    PerformerEditComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    PerformerEditComponent.prototype.localePipe;
}
