// @gsp-wf/wf-process-editor/esm5/lib/views/performer/views/matrix-performer/enum-list/enum-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, HostBinding, ViewChild, TemplateRef, Injector, Optional, ChangeDetectorRef, LOCALE_ID, Inject } from '@angular/core';
import { DatagridComponent } from '@farris/ui-datagrid';
import { Subject } from 'rxjs/Subject';
import { MessagerService } from '@farris/ui-messager';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../../../../i18n/pipe/wf-process-editor-locale.pipe';
var EnumListComponent = /** @class */ (function () {
    function EnumListComponent(messageService, injector, cdr, localeId, resources, localePipe) {
        this.messageService = messageService;
        this.injector = injector;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.cls = 'farris-main-area flex-column';
        this.columns = [
            { field: 'key', width: 100, title: this.localePipe.transform('component.matrixPerformer.key') },
            { field: 'value', width: 100, title: this.localePipe.transform('component.matrixPerformer.value') }
        ];
        this.confirm = new Subject();
        this.closeModalEmitter = new Subject();
        /**
         * 显示值
         */
        this.data = new Array();
        this.selectedIds = new Array();
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
    }
    /**
     * @return {?}
     */
    EnumListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * 加载显示数据
     */
    /**
     * 加载显示数据
     * @param {?} enumValues
     * @return {?}
     */
    EnumListComponent.prototype.load = /**
     * 加载显示数据
     * @param {?} enumValues
     * @return {?}
     */
    function (enumValues) {
        this.data = tslib_1.__spread(enumValues);
    };
    /**
     * 确认
     */
    /**
     * 确认
     * @return {?}
     */
    EnumListComponent.prototype.onConfirm = /**
     * 确认
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.selectedIds.length === 0) {
            this.messageService.warning(this.localePipe.transform('component.matrixPerformer.notSelectEnum'));
        }
        /** @type {?} */
        var enumValues = this.data.filter((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return _this.selectedIds.findIndex((/**
         * @param {?} id
         * @return {?}
         */
        function (id) { return id === d.key; })) > -1; }));
        this.confirm.next(enumValues);
        this.closeModalEmitter.next();
    };
    /**
     * 取消
     */
    /**
     * 取消
     * @return {?}
     */
    EnumListComponent.prototype.cancel = /**
     * 取消
     * @return {?}
     */
    function () {
        this.closeModalEmitter.next();
    };
    EnumListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-enum-list',
                    template: "<ng-template #buttons>\r\n  <button class=\"btn btn-secondary\" (click)=\"cancel()\">{{'component.matrixPerformer.cancel'|wfLocale}}</button>\r\n  <button class=\"btn btn-primary\" (click)=\"onConfirm()\">{{'component.matrixPerformer.onConfirm'|wfLocale}}</button>\r\n</ng-template>\r\n\r\n<div class=\"f-page\">\r\n  <div class=\"f-page-main f-utils-fill-flex-column h-100\">\r\n    <div class=\"f-utils-fill-flex-column farris-overflow-y-auto h-100\">\r\n      <farris-datagrid [data]=\"data\" [columns]=\"columns\" #dg=\"datagrid\" [fit]=\"true\"\r\n                       [idField]=\"'key'\" [fitColumns]=\"true\" [showLineNumber]=\"true\" [(checkValues)]=\"selectedIds\"\r\n                       [showAllCheckbox]=\"true\" [showCheckbox]=\"true\" [multiSelect]=\"false\" [checkOnSelect]=\"true\" [selectOnCheck]=\"true\"\r\n                       lineNumberTitle=\"{{'component.matrixPerformer.lineNumber'|wfLocale}}\" [lineNumberWidth]=\"50\" [striped]=\"true\" [virtualized]=\"true\" [pagination]=\"false\">\r\n      </farris-datagrid>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n",
                    providers: [
                        WfLocalePipe,
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    EnumListComponent.ctorParameters = function () { return [
        { type: MessagerService },
        { type: Injector },
        { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
        { type: WfLocalePipe, decorators: [{ type: Optional }] }
    ]; };
    EnumListComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        dg: [{ type: ViewChild, args: ['dg',] }],
        btnRef: [{ type: ViewChild, args: ['buttons',] }]
    };
    return EnumListComponent;
}());
export { EnumListComponent };
if (false) {
    /** @type {?} */
    EnumListComponent.prototype.cls;
    /** @type {?} */
    EnumListComponent.prototype.columns;
    /** @type {?} */
    EnumListComponent.prototype.dg;
    /** @type {?} */
    EnumListComponent.prototype.btnRef;
    /** @type {?} */
    EnumListComponent.prototype.confirm;
    /** @type {?} */
    EnumListComponent.prototype.closeModalEmitter;
    /**
     * 显示值
     * @type {?}
     */
    EnumListComponent.prototype.data;
    /** @type {?} */
    EnumListComponent.prototype.selectedIds;
    /**
     * @type {?}
     * @private
     */
    EnumListComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    EnumListComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    EnumListComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    EnumListComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    EnumListComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    EnumListComponent.prototype.localePipe;
}
