// @farris/component-querycondition/esm2015/lib/directives/label-text-multi.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/label-text-multi.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Injector, ElementRef, NgZone } from '@angular/core';
export class LabelTextMultiDirective {
    /**
     * @param {?} el
     * @param {?} injector
     */
    constructor(el, injector) {
        this.el = el;
        this.injector = injector;
        this.multiCls = 'farris-group-multi-label';
        this._enableMulti = false;
        // 获取ngZone
        this.ngZone = null;
        if (!this.ngZone) {
            this.ngZone = this.injector.get(NgZone, null);
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set enableMulti(value) {
        this._enableMulti = value;
        this.judgeMulti();
    }
    /**
     * @return {?}
     */
    get enableMulti() {
        return this._enableMulti;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    judgeMulti() {
        if (this._enableMulti) {
            this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => {
                setTimeout((/**
                 * @return {?}
                 */
                () => {
                    /** @type {?} */
                    let textEl = this.el.nativeElement.querySelector('.farris-label-text');
                    if (!textEl) {
                        return;
                    }
                    // 超出
                    if (textEl.clientWidth < textEl.scrollWidth) {
                        this.el.nativeElement.className = this.el.nativeElement.className + ' ' + this.multiCls;
                    }
                    else {
                        this.el.nativeElement.className = this.el.nativeElement.className.replace(this.multiCls, '');
                    }
                }), 20);
            }));
        }
    }
}
LabelTextMultiDirective.decorators = [
    { type: Directive, args: [{
                selector: '[label-text-multi]'
            },] }
];
/** @nocollapse */
LabelTextMultiDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Injector }
];
LabelTextMultiDirective.propDecorators = {
    enableMulti: [{ type: Input, args: ['label-text-multi',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    LabelTextMultiDirective.prototype.multiCls;
    /**
     * @type {?}
     * @private
     */
    LabelTextMultiDirective.prototype._enableMulti;
    /**
     * @type {?}
     * @private
     */
    LabelTextMultiDirective.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    LabelTextMultiDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    LabelTextMultiDirective.prototype.injector;
}
