// @farris/command-services/esm5/lib/languages/language-message-zhft.js
/**
 * 繁体中文语言消息
 */
var LanguageMessageForZhft = /** @class */ (function () {
    function LanguageMessageForZhft() {
        this.yes = '是';
        this.no = '否';
        this.unsaveNotifyTitle = '存在未保存記錄。是否要保存？';
        this.clientNotifyTitle = '前端異常提示';
        this.serverNotifyTitle = '服務器端異常提示';
        this.cancelApproveSuccess = '取消提交成功！';
        this.cancelApproveFailed = '取消提交失敗！';
        this.unallowEmptyProcessInstanceName = '流程實例名稱不能為空';
        this.unallowEmptyBizBillId = '請選擇壹條數據';
        this.bizDefKeyRequired = '入口單據不能為空';
        this.unallowEmptyChildBizBillId = '從表業務單據編號不能為空！';
        this.procInsIdRequired = '流程實例id不能為空！';
        this.addChildFailed = '新增子級失敗';
        this.addSiblingFailed = '新增同級失敗';
        this.addSubChildFailed = '新增從表子級失敗';
        this.addSubSiblingFailed = '新增從表同級失敗';
        this.deleteFailed = '刪除失敗';
        this.multiSaveFailed = '批量保存失敗';
        this.appendFailed = '追加失敗';
        this.queryFailed = '查詢失敗';
        this.cancelFailed = '取消失敗';
        this.updateFailed = '更新失敗';
        this.addFailed = '新增失敗';
        this.loadFailed = '加載失敗';
        this.saveSuccess = '保存成功！';
        this.saveFailed = '保存失敗！';
        this.deleteSuccess = '刪除成功！';
        this.deleteFaild = '刪除失敗！';
        this.confirmDeletion = '確認刪除？';
        this.submitSuccess = '提交審批成功！';
        this.submitFaild = '提交審批失敗！';
        this.notifyTitle = '系統提示';
        this.httpError = 'HTTP請求錯誤！請檢查Server端錯誤！';
        this.gridDataNotSave = '當前頁數據未保存，翻頁將導致數據丟失，是否繼續翻頁？';
        this.exitWithoutSave = '存在未保存的數據，是否繼續關閉？';
        this.notSupportMenuType = '關閉的既不是菜單也不是應用。';
        this.cancelWithoutSave = '存在未保存的變更，確認取消？';
        this.plsSelectDeleteData = '請選擇要刪除的數據！';
        this.errorHierarchyKey = '錯誤的分級碼';
        this.plsSelectParentNode = '請選擇父節點';
        this.deleteChildFirst = '請先刪除子節點';
        this.incorrectHierarchyConfig = '分級碼配置信息錯誤';
        this.operationFailed = '操作執行失敗。';
        this.plsSelectEditData = '請選擇要編輯的數據！';
        this.plsSelectViewData = '請選擇要查看的數據！';
        this.plsUploadFirst = '請先上傳附件！';
        this.defaultDialogTitle = '彈窗';
        this.changeToFirst = '已到達第壹條數據';
        this.changeToLast = '已到達最後壹條數據';
        this.noProcessInstanceId = '請指定流程實例標識。';
        this.noDataExist = '要編輯的數據不存在，無法進入編輯狀態！';
        this.noAttachment = '沒有可以預覽的附件。';
        this.confirm = '確定';
        this.cancel = '取消';
        this.plsSelectCopyData = '請選擇要復制的數據！';
        this.copyFieldsRequired = '要復制的字段不能為空！';
        this.pathIsRequired = '請求路径不能為空！';
        this.propsIsEmpty = '沒有可以編輯的物料特征！';
        this.historyAttachment = '曆史版本附件禁止刪除！';
        this.plsSelectDownloadAtt = '請選擇要下載的附件!';
        this.noDownloadAtt = '找不到要下載的附件！';
        this.plsCheckBatchEditRows = '請勾選要批量編輯的行！';
        this.plsSelectDetailFormData = '請先選擇一條從表數據！';
        this.dataAndStateChanged = '確定要加載數據並切換到$1狀態？';
        this.dataChanged = '確定要加載數據並放棄當前修改？';
        this.stateChanged = '確定要切換到$1狀態？';
        this.defaultStateName = '初始';
        this.copy = '復制詳細信息';
        this.copySuccess = '復制成功';
        this.copyFailed = '復制失敗';
        this.roger = '知道了';
        this.appOrFuncIdRequired = '未配置菜單或應用參數，請在設計器中配置。';
        this.tableCanNotEmpty = '不能為空。';
        this.plsSelectRemoveAtt = '請選擇要刪除的附件!';
        this.plsSelectUpdateRow = '請選擇更新附件的行！';
    }
    return LanguageMessageForZhft;
}());
export { LanguageMessageForZhft };
