// @farris/command-services/esm5/lib/languages/language-message-zh.js
/**
 * 英文语言消息
 */
var LanguageMessageForZh = /** @class */ (function () {
    function LanguageMessageForZh() {
        this.yes = '是';
        this.no = '否';
        this.unsaveNotifyTitle = '存在未保存记录。是否要保存？';
        this.clientNotifyTitle = '前端异常提示';
        this.serverNotifyTitle = '服务器端异常提示';
        this.cancelApproveSuccess = '取消提交成功！';
        this.cancelApproveFailed = '取消提交失败！';
        this.unallowEmptyProcessInstanceName = '流程实例名称不能为空';
        this.unallowEmptyBizBillId = '请选择一条数据';
        this.bizDefKeyRequired = '入口单据不能为空';
        this.unallowEmptyChildBizBillId = '从表业务单据编号不能为空！';
        this.procInsIdRequired = '流程实例Id不能为空！';
        this.addChildFailed = '新增子级失败';
        this.addSiblingFailed = '新增同级失败';
        this.addSubChildFailed = '新增从表子级失败';
        this.addSubSiblingFailed = '新增从表同级失败';
        this.deleteFailed = '删除失败';
        this.multiSaveFailed = '批量保存失败';
        this.appendFailed = '追加失败';
        this.queryFailed = '查询失败';
        this.cancelFailed = '取消失败';
        this.updateFailed = '更新失败';
        this.addFailed = '新增失败';
        this.loadFailed = '加载失败';
        this.saveSuccess = '保存成功！';
        this.saveFailed = '保存失败！';
        this.deleteSuccess = '删除成功！';
        this.deleteFaild = '删除失败！';
        this.confirmDeletion = '确认删除？';
        this.submitSuccess = '提交审批成功！';
        this.submitFaild = '提交审批失败！';
        this.notifyTitle = '系统提示';
        this.httpError = 'HTTP请求错误！请检查Server端错误！';
        this.gridDataNotSave = '当前页数据未保存，翻页将导致数据丢失，是否继续翻页？';
        this.exitWithoutSave = '存在未保存的数据，是否继续关闭？';
        this.notSupportMenuType = '关闭的既不是菜单也不是应用。';
        this.cancelWithoutSave = '存在未保存的变更，确认取消？';
        this.plsSelectDeleteData = '请选择要删除的数据！';
        this.errorHierarchyKey = '错误的分级码';
        this.plsSelectParentNode = '请选择父节点';
        this.deleteChildFirst = '请先删除子节点';
        this.incorrectHierarchyConfig = '分级码配置信息错误';
        this.operationFailed = '操作执行失败。';
        this.plsSelectEditData = '请选择要编辑的数据！';
        this.plsSelectViewData = '请选择要查看的数据！';
        this.plsUploadFirst = '请先上传附件！';
        this.defaultDialogTitle = '弹窗';
        this.changeToFirst = '已到达第一条数据';
        this.changeToLast = '已到达最后一条数据';
        this.noProcessInstanceId = '请指定流程实例标识。';
        this.noDataExist = '要编辑的数据不存在，无法进入编辑状态！';
        this.noAttachment = '没有可以预览的附件。';
        this.confirm = '确定';
        this.cancel = '取消';
        this.plsSelectCopyData = '请选择要复制的数据！';
        this.copyFieldsRequired = '要复制的字段不能为空！';
        this.pathIsRequired = '请求路径不能为空！';
        this.propsIsEmpty = '没有可以编辑的物料特征！';
        this.historyAttachment = '历史版本附件禁止删除！';
        this.plsSelectDownloadAtt = '请选择要下载的附件!';
        this.noDownloadAtt = '找不到要下载的附件!';
        this.plsCheckBatchEditRows = '请勾选要批量编辑的行！';
        this.plsSelectDetailFormData = '请先选择一条从表数据！';
        this.dataAndStateChanged = '确定要加载数据并切换到$1状态？';
        this.dataChanged = '确定要加载数据并放弃当前修改？';
        this.stateChanged = '确定要切换到$1状态？';
        this.defaultStateName = '初始';
        this.copy = '复制详细信息';
        this.copySuccess = '复制成功';
        this.copyFailed = '复制失败';
        this.roger = '知道了';
        this.appOrFuncIdRequired = '未配置菜单或应用参数，请在设计器中配置。';
        this.tableCanNotEmpty = '不能为空。';
        this.plsSelectRemoveAtt = '请选择要删除的附件!';
        this.plsSelectUpdateRow = '请选择要更新附件的行！';
    }
    return LanguageMessageForZh;
}());
export { LanguageMessageForZh };
