// @farris/mobile-vue-adapter/esm2015/lib/attachment/attachment-filter.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/attachment/attachment-filter.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { getDataByPath, getAttachmentUrl } from '../utils';
/** @type {?} */
export const filterData = (/**
 * @param {?} rawData
 * @param {?} field
 * @param {?} rootId
 * @param {?=} isComponentAdapter
 * @return {?}
 */
(rawData, field, rootId, isComponentAdapter = true) => {
    /** @type {?} */
    const splitFiled = field.split('/');
    /** @type {?} */
    const fields = splitFiled.filter((/**
     * @param {?} ele
     * @return {?}
     */
    ele => ele));
    if (!isComponentAdapter) {
        // 单条数据
        /** @type {?} */
        const data = getDataByPath(rawData, fields);
        /** @type {?} */
        let url = '/platform/common/web/assets/imgs/no-pic.png';
        // 附件不存在
        if (!data || !data.attachmentId) {
            return url;
        }
        // 附件存在
        /** @type {?} */
        const extensionName = data.fileName.substring(data.fileName.lastIndexOf('.') + 1);
        url = getAttachmentUrl(data.attachmentId, rootId, extensionName);
        return url;
    }
    /** @type {?} */
    const transformData = rawData.map((/**
     * @param {?} everyData
     * @return {?}
     */
    everyData => {
        /** @type {?} */
        const data = getDataByPath(everyData, fields);
        /** @type {?} */
        let url = '/platform/common/web/assets/imgs/no-pic.png';
        if (!data || !data.attachmentId) {
            return {
                url,
                id: everyData.id,
                attachmentId: data.attachmentId,
                name: '',
                size: 0,
                createTime: ''
            };
        }
        /** @type {?} */
        const extensionName = data.fileName.substring(data.fileName.lastIndexOf('.') + 1);
        url = getAttachmentUrl(data.attachmentId, rootId, extensionName);
        return {
            url,
            id: everyData.id,
            attachmentId: data.attachmentId,
            name: data.fileName,
            size: data.fileSize,
            createTime: data.fileCreateTime
        };
    }));
    return transformData;
});
