// @farris/mobile-vue-adapter/esm2015/lib/attachment/attachment.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/attachment/attachment.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
/**
 * @abstract
 */
export class AttachmentService {
    constructor() {
        this.collectFile = new Subject();
    }
    /**
     * 拍照
     * @param {?=} param 拍照参数
     * @return {?}
     */
    takePhoto(param) {
        this.takePhotoIns.takePhotoAttachment(param);
    }
    /**
     * 选择文件
     * @param {?} param 选择文件参数
     * @return {?}
     */
    selectAttachment(param) {
        this.takePhotoIns.selectAttachments(param);
    }
    /**
     * 拍摄短视频
     * @param {?} param 拍摄短视频参数
     * @return {?}
     */
    takeShortVideo(param) {
        this.takePhotoIns.takeShortVideo(param);
    }
    /**
     * 播放短视频
     * @param {?} param 播放短视频参数
     * @return {?}
     */
    playShortVideo(param) {
        this.takePhotoIns.playShortVideo(param);
    }
    /**
     * @param {?} param
     * @return {?}
     */
    selectFile(param) {
        this.takePhotoIns.selectFile(param);
    }
    /**
     * @param {?} param
     * @return {?}
     */
    downloadAttachment(param) {
        this.takePhotoIns.downloadAttachment(param);
    }
    /**
     * @return {?}
     */
    getSysInfo() {
        return this.takePhotoIns.getSysInfo();
    }
}
if (false) {
    /** @type {?} */
    AttachmentService.prototype.takePhotoIns;
    /** @type {?} */
    AttachmentService.prototype.collectFile;
}
