// @farris/ui-responsive/esm2015/lib/container/res-container.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, HostBinding, Renderer2 } from '@angular/core';
export class FarrisContainerDirective {
    /**
     * @param {?} el
     * @param {?} render
     */
    constructor(el, render) {
        this.el = el;
        this.render = render;
        this.isContainer = true;
    }
    /**
     * @return {?}
     */
    get width() {
        // 对各个单位进行处理
        return this._width;
    }
    /**
     * @param {?} width
     * @return {?}
     */
    set width(width) {
        this._width = this.resolveSize(width);
    }
    /**
     * @return {?}
     */
    get height() {
        return this._height;
    }
    /**
     * @param {?} height
     * @return {?}
     */
    set height(height) {
        this._height = this.resolveSize(height);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // 宽高  赋值
        this.render.setStyle(this.el.nativeElement, 'width', this.width);
        this.render.setStyle(this.el.nativeElement, 'height', this.height);
    }
    /*宽高类型是 string或者number 解析宽高  尺寸 */
    /**
     * @param {?} size
     * @return {?}
     */
    resolveSize(size) {
        /** @type {?} */
        const regex = /px|em|rem|pt|%/;
        // 说明是字符串
        return regex.test(size) ? `${parseInt(size, 10)}${size.match(regex)[0]}` : `${size}px`;
    }
}
FarrisContainerDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-container]'
            },] }
];
/** @nocollapse */
FarrisContainerDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
FarrisContainerDirective.propDecorators = {
    isContainer: [{ type: HostBinding, args: ['class.container',] }],
    width: [{ type: Input }],
    height: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FarrisContainerDirective.prototype.isContainer;
    /** @type {?} */
    FarrisContainerDirective.prototype._width;
    /** @type {?} */
    FarrisContainerDirective.prototype._height;
    /**
     * @type {?}
     * @private
     */
    FarrisContainerDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FarrisContainerDirective.prototype.render;
}
