// @qdp/formular/esm2015/lib/component/formular-define-wizard/formular-define-wizard.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component/formular-define-wizard/formular-define-wizard.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild } from '@angular/core';
import { FormularSelectComponent } from '../formular-select/formular-select.component';
import { CacheService } from '@ecp-caf/caf-common';
import { FormularDefineComponent } from '../formular-define/formular-define.component';
import { StepComponentComponent } from '../step-component/step-component.component';
import { EntityDataService } from '../../service/entitydata.service';
import { LocalizeService } from '@qdp/localize';
export class FormularDefineWizardComponent {
    /**
     * @param {?} cacheService
     * @param {?} entityDataService
     * @param {?} localizeService
     */
    constructor(cacheService, entityDataService, localizeService) {
        this.cacheService = cacheService;
        this.entityDataService = entityDataService;
        this.localizeService = localizeService;
        this.steps = [
            {
                stepId: 'selectTemplate',
                title: this.localizeService.getValue('formular.defineWizard.labelSelectTemplate'),
                component: FormularSelectComponent,
                data: '',
                show: true,
                nextStep: 'formularDefine',
                showFinishBtn: true,
                showCancelBtn: true,
                showLastStepBtn: false,
                showNextStepBtn: true
            },
            {
                stepId: 'formularDefine',
                title: this.localizeService.getValue('formular.defineWizard.labelFormularDefine'),
                component: FormularDefineComponent,
                data: '',
                show: true,
                nextStep: '',
                showFinishBtn: true,
                showCancelBtn: true,
                showLastStepBtn: true,
                showNextStepBtn: false
            }
        ];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.cacheService.set('formularParam', { currentSU: this.currentSU, currentFormularId: this.currentFormularId, serverHost: this.serverHost });
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.entityDataService.setFormularState(false, {});
        if (this.currentFormularId) {
            /** @type {?} */
            const formularEdit = {
                stepId: 'formularDefine',
                title: this.localizeService.getValue('formular.defineWizard.labelFormularDefine'),
                component: FormularDefineComponent,
                data: '',
                show: true,
                nextStep: '',
                showFinishBtn: true,
                showCancelBtn: true,
                showLastStepBtn: true,
                showNextStepBtn: false
            };
            this.qdpsteps.jumpStep(formularEdit);
        }
    }
    /**
     * 点击完成
     * @return {?}
     */
    clickFinishBtn() {
        if (this.parentDialog) {
            // this.selectFormularInfo = this.cacheService.get('formularResult');
            this.parentDialog.close();
        }
    }
}
FormularDefineWizardComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-formular-define-wizard',
                template: "<lib-step-component #qdpsteps [steps]=\"steps\" (clickFinishBtn)=\"clickFinishBtn()\">\r\n</lib-step-component>",
                styles: [""]
            }] }
];
/** @nocollapse */
FormularDefineWizardComponent.ctorParameters = () => [
    { type: CacheService },
    { type: EntityDataService },
    { type: LocalizeService }
];
FormularDefineWizardComponent.propDecorators = {
    qdpsteps: [{ type: ViewChild, args: ['qdpsteps',] }]
};
if (false) {
    /** @type {?} */
    FormularDefineWizardComponent.prototype.qdpsteps;
    /** @type {?} */
    FormularDefineWizardComponent.prototype.currentSU;
    /** @type {?} */
    FormularDefineWizardComponent.prototype.currentFormularId;
    /** @type {?} */
    FormularDefineWizardComponent.prototype.currentFunctionId;
    /** @type {?} */
    FormularDefineWizardComponent.prototype.serverHost;
    /** @type {?} */
    FormularDefineWizardComponent.prototype.sessionId;
    /** @type {?} */
    FormularDefineWizardComponent.prototype.parentDialog;
    /** @type {?} */
    FormularDefineWizardComponent.prototype.formularType;
    /** @type {?} */
    FormularDefineWizardComponent.prototype.selectFormularInfo;
    /** @type {?} */
    FormularDefineWizardComponent.prototype.steps;
    /**
     * @type {?}
     * @private
     */
    FormularDefineWizardComponent.prototype.cacheService;
    /**
     * @type {?}
     * @private
     */
    FormularDefineWizardComponent.prototype.entityDataService;
    /**
     * @type {?}
     * @private
     */
    FormularDefineWizardComponent.prototype.localizeService;
}
