// @farris/ui-forms/esm5/lib/checkbox/checkbox-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { IdService } from '@farris/ui-common';
var CheckboxGroupComponent = /** @class */ (function () {
    function CheckboxGroupComponent(idSer) {
        this.idSer = idSer;
        /* 初始化 checkbox 的值 */
        this.model = [];
        /*  */
        this.modelChange = new EventEmitter();
        this.controlChange = (/**
         * @return {?}
         */
        function () { });
        this.controlTouch = (/**
         * @return {?}
         */
        function () { });
    }
    /**
     * @return {?}
     */
    CheckboxGroupComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.name = this.idSer.generate();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    CheckboxGroupComponent.prototype.changeModel = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var index = this.model.findIndex((/**
         * @param {?} ele
         * @return {?}
         */
        function (ele) {
            return ele === value;
        }));
        // 如果已经选中  则设置为不选中
        if (index > -1) {
            this.model.splice(index, 1);
            return this.modelChange.emit(this.model);
        }
        // 如果没有选中  设置为选中
        this.model.push(value);
        // control的值设置为逗号分隔的字符串  便于存储和传递
        this.controlChange(this.model.join());
        this.modelChange.emit(this.model);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    CheckboxGroupComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.model = value;
        this.changeModel(this.model);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    CheckboxGroupComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.controlChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    CheckboxGroupComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.controlTouch = fn;
    };
    CheckboxGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-checkbox-group',
                    template: "\n    <div class=\"farris-input-wrap\" [class.farris-checkradio-hor]=\"horizontal\">\n      <ng-content></ng-content>\n    </div>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return CheckboxGroupComponent; })),
                            multi: true,
                        }]
                }] }
    ];
    /** @nocollapse */
    CheckboxGroupComponent.ctorParameters = function () { return [
        { type: IdService }
    ]; };
    CheckboxGroupComponent.propDecorators = {
        model: [{ type: Input }],
        name: [{ type: Input }],
        horizontal: [{ type: Input }],
        modelChange: [{ type: Output }]
    };
    return CheckboxGroupComponent;
}());
export { CheckboxGroupComponent };
if (false) {
    /** @type {?} */
    CheckboxGroupComponent.prototype.model;
    /** @type {?} */
    CheckboxGroupComponent.prototype.name;
    /** @type {?} */
    CheckboxGroupComponent.prototype.horizontal;
    /** @type {?} */
    CheckboxGroupComponent.prototype.modelChange;
    /**
     * @type {?}
     * @private
     */
    CheckboxGroupComponent.prototype.controlChange;
    /**
     * @type {?}
     * @private
     */
    CheckboxGroupComponent.prototype.controlTouch;
    /**
     * @type {?}
     * @private
     */
    CheckboxGroupComponent.prototype.idSer;
}
