// @farris/ui-forms/esm5/lib/checkbox/checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ElementRef, forwardRef, Output, EventEmitter } from '@angular/core';
import { CheckboxGroupComponent } from './checkbox-group.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var CheckboxComponent = /** @class */ (function () {
    function CheckboxComponent(checkbox, el) {
        this.checkbox = checkbox;
        this.el = el;
        this.modelChange = new EventEmitter();
        this.controlChange = (/**
         * @return {?}
         */
        function () { });
        this.controlTouch = (/**
         * @return {?}
         */
        function () { });
    }
    Object.defineProperty(CheckboxComponent.prototype, "name", {
        /* radio 原生name 用于多个radio单选使用 */
        get: /* radio 原生name 用于多个radio单选使用 */
        /**
         * @return {?}
         */
        function () {
            if (this.hasParent(this.el.nativeElement)) {
                return this.checkbox.name;
            }
            return '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckboxComponent.prototype, "model", {
        /* model */
        get: /* model */
        /**
         * @return {?}
         */
        function () {
            var _this = this;
            if (this.hasParent(this.el.nativeElement)) {
                return this.checkbox.model.find((/**
                 * @param {?} ele
                 * @return {?}
                 */
                function (ele) {
                    return ele === _this.label;
                }));
            }
            return this.nyModel;
        },
        set: /**
         * @param {?} model
         * @return {?}
         */
        function (model) {
            this.nyModel = model;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    CheckboxComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /* 值传递 */
    /* 值传递 */
    /**
     * @return {?}
     */
    CheckboxComponent.prototype.changeModelHandler = /* 值传递 */
    /**
     * @return {?}
     */
    function () {
        // 父元素获取label  子元素再从父元素获取model
        if (this.hasParent(this.el.nativeElement)) {
            return this.checkbox.changeModel(this.label);
        }
        this.model = this.label;
        this.modelChange.emit(this.model);
        this.controlChange(this.label);
    };
    /* 是否存在父元素  存在即单选组*/
    /* 是否存在父元素  存在即单选组*/
    /**
     * @param {?} element
     * @return {?}
     */
    CheckboxComponent.prototype.hasParent = /* 是否存在父元素  存在即单选组*/
    /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        return element.parentElement.classList.contains('farris-input-wrap');
    };
    /**
     * @param {?} value
     * @return {?}
     */
    CheckboxComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.model = value;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    CheckboxComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.controlChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    CheckboxComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.controlTouch = fn;
    };
    CheckboxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-checkbox',
                    template: "<label class=\"custom-control custom-checkbox\">\r\n  <input class=\"custom-control-input\" [name]=\"name\" [value]=\"label\" [ngModel]=\"model\"\r\n    (ngModelChange)=\"changeModelHandler()\" type=\"checkbox\" [disabled]=\"disabled\">\r\n  <span class=\"custom-control-label\">\r\n    <ng-content>\r\n    </ng-content>\r\n  </span>\r\n</label>",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return CheckboxComponent; })),
                            multi: true,
                        }],
                    styles: ['./checkbox.component.css']
                }] }
    ];
    /** @nocollapse */
    CheckboxComponent.ctorParameters = function () { return [
        { type: CheckboxGroupComponent },
        { type: ElementRef }
    ]; };
    CheckboxComponent.propDecorators = {
        model: [{ type: Input }],
        label: [{ type: Input }],
        disabled: [{ type: Input }],
        modelChange: [{ type: Output }]
    };
    return CheckboxComponent;
}());
export { CheckboxComponent };
if (false) {
    /** @type {?} */
    CheckboxComponent.prototype.label;
    /** @type {?} */
    CheckboxComponent.prototype.disabled;
    /** @type {?} */
    CheckboxComponent.prototype.modelChange;
    /** @type {?} */
    CheckboxComponent.prototype.nyModel;
    /**
     * @type {?}
     * @private
     */
    CheckboxComponent.prototype.controlChange;
    /**
     * @type {?}
     * @private
     */
    CheckboxComponent.prototype.controlTouch;
    /**
     * @type {?}
     * @private
     */
    CheckboxComponent.prototype.checkbox;
    /**
     * @type {?}
     * @private
     */
    CheckboxComponent.prototype.el;
}
