// @gsp-lcm/bo-creator/esm5/lib/views/bo-creator/modal/app-modal/app-creator-modal.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/views/bo-creator/modal/app-modal/app-creator-modal.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter } from '@angular/core';
import { NotifyService } from '@farris/ui-notify';
import { BusinessObjectEntity, BusinessObjectDtService } from '@gsp-lcm/bo-dt-service';
import { ValidatorResult } from '../../common/validation-result';
import { BoCreatorUtils } from '../../utils/bo-create-utils';
var AppCreatorModalComponent = /** @class */ (function () {
    function AppCreatorModalComponent(boCreatorUtils, boDtService, notifyService) {
        this.boCreatorUtils = boCreatorUtils;
        this.boDtService = boDtService;
        this.notifyService = notifyService;
        this.appConfirmEmitter = new EventEmitter();
        this.currentBo = new BusinessObjectEntity();
        this.apps = [];
    }
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initData();
    };
    /**
     * @private
     * @return {?}
     */
    AppCreatorModalComponent.prototype.initData = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.boDtService.getBusinessObjects().subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (!data) {
                return;
            }
            //二级节点是APP
            _this.apps = data.filter((/**
             * @param {?} x
             * @return {?}
             */
            function (x) { return x.layer === 2; }));
            //一级节点是产品，目前默认一级节点（根节点）只有一个
            _this.parent = data.find((/**
             * @param {?} x
             * @return {?}
             */
            function (x) { return x.layer === 1; }));
        }));
    };
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.create = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var bo = this.getCurrentBo();
        if (!bo) {
            return;
        }
        this.boDtService.save(bo).subscribe((/**
         * @return {?}
         */
        function () {
            _this.appConfirmEmitter.emit(bo);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.notifyService.error('创建失败');
            throw error;
        }));
    };
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.getCurrentBo = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = this.appInfoValidator();
        if (result.illegal) {
            this.notifyService.warning(result.message);
            return null;
        }
        this.currentBo.id = this.boCreatorUtils.newGuid();
        this.currentBo.layer = 2;
        this.currentBo.parentID = this.parent.id;
        this.currentBo.sysInit = '0';
        this.currentBo.isSysInit = false;
        this.currentBo.isDetail = '0';
        /** @type {?} */
        var max = this.boCreatorUtils.getMaxOrder(this.apps, this.currentBo.parentID);
        this.currentBo.sortOrder = max + 1;
        return Object.assign(new BusinessObjectEntity(), this.currentBo);
    };
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.appInfoValidator = /**
     * @return {?}
     */
    function () {
        //新增信息空校验
        /** @type {?} */
        var result = this.emptyValidator();
        if (result.illegal) {
            return result;
        }
        //编号，名称命名规范性校验
        result = this.namingConventionValidator();
        if (result.illegal) {
            return result;
        }
        //APP编号重复校验
        return this.repeatValidator();
    };
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.emptyValidator = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = new ValidatorResult();
        if (!this.currentBo.code) {
            result.illegal = true;
            result.message = "\u5173\u952E\u5E94\u7528\u7684\u7F16\u53F7\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        if (!this.currentBo.name) {
            result.illegal = true;
            result.message = "\u5173\u952E\u5E94\u7528\u7684\u540D\u79F0\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        return result;
    };
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.namingConventionValidator = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = new ValidatorResult();
        /** @type {?} */
        var codeRegx = new RegExp('^[a-zA-Z\x7f-\xff][a-zA-Z0-9\x7f-\xff]*$');
        if (!codeRegx.test(this.currentBo.code)) {
            result.illegal = true;
            result.message = '关键应用的编号只能包含字母,数字,并且不可以用数字开头!';
            return result;
        }
        /** @type {?} */
        var nameRegx = new RegExp('[`~!@#$^&*()=|{}\':;\',\\[\\].<>/?~！@#￥……&*（）——|{}【】‘；：”“\'。，、？]');
        if (nameRegx.test(this.currentBo.name)) {
            result.illegal = true;
            result.message = '关键应用的名称包含非法字符，请重新输入!';
            return result;
        }
        return result;
    };
    /**
     * @return {?}
     */
    AppCreatorModalComponent.prototype.repeatValidator = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var result = new ValidatorResult();
        /** @type {?} */
        var filterApp = this.apps.find((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.code.toLowerCase() === _this.currentBo.code.toLowerCase(); }));
        if (filterApp) {
            result.illegal = true;
            result.message = "\u5173\u952E\u5E94\u7528\u7684\u7F16\u53F7" + this.currentBo.code + "\u5DF2\u5B58\u5728";
            return result;
        }
        return result;
    };
    AppCreatorModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-app-creator-modal',
                    template: "\r\n<div class=\"cloumn\" style=\"margin-top: 28px;\">\r\n\r\n  <div class=\"app-Info row\">\r\n    <label class=\"col-form-label ps-col-form-label\">\r\n      <span class=\"farris-label-info text-danger\">*</span>\r\n      <span class=\"farris-label-text\">\u5E94\u7528\u7F16\u53F7</span>\r\n    </label>\r\n    <div class=\"col-9\">\r\n      <input-group [(ngModel)]=\"currentBo.code\"  autocomplete=\"off\" name=\"code\" ></input-group>\r\n    </div>\r\n  </div>\r\n\r\n  <div class=\"app-Info row\">\r\n    <label class=\"col-form-label ps-col-form-label\">\r\n      <span class=\"farris-label-info text-danger\">*</span>\r\n      <span class=\"farris-label-text\">\u5E94\u7528\u540D\u79F0</span>\r\n    </label>\r\n    <div class=\"col-9\">\r\n      <input-group [(ngModel)]=\"currentBo.name\"  autocomplete=\"off\" name=\"name\" ></input-group>\r\n    </div>\r\n  </div>\r\n\r\n  <div class=\"app-Info row\">\r\n    <label class=\"col-form-label ps-col-form-label\">\r\n      <span class=\"farris-label-text\">\u5E94\u7528\u63CF\u8FF0</span>\r\n    </label>\r\n    <div class=\"col-9\">\r\n      <input-group [(ngModel)]=\"currentBo.description\"  autocomplete=\"off\" name=\"description\"></input-group>\r\n    </div>\r\n  </div>\r\n\r\n</div>",
                    styles: [".app-Info{margin:0 0 16px 20px}div .ps-col-form-label{width:94px;text-align:right}"]
                }] }
    ];
    /** @nocollapse */
    AppCreatorModalComponent.ctorParameters = function () { return [
        { type: BoCreatorUtils },
        { type: BusinessObjectDtService },
        { type: NotifyService }
    ]; };
    AppCreatorModalComponent.propDecorators = {
        appConfirmEmitter: [{ type: Output }]
    };
    return AppCreatorModalComponent;
}());
export { AppCreatorModalComponent };
if (false) {
    /** @type {?} */
    AppCreatorModalComponent.prototype.appConfirmEmitter;
    /** @type {?} */
    AppCreatorModalComponent.prototype.currentBo;
    /** @type {?} */
    AppCreatorModalComponent.prototype.parent;
    /** @type {?} */
    AppCreatorModalComponent.prototype.apps;
    /**
     * @type {?}
     * @private
     */
    AppCreatorModalComponent.prototype.boCreatorUtils;
    /**
     * @type {?}
     * @private
     */
    AppCreatorModalComponent.prototype.boDtService;
    /**
     * @type {?}
     * @private
     */
    AppCreatorModalComponent.prototype.notifyService;
}
