// @farris/ui-locale/esm5/lib/locale.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FarrisCommonModule } from '@farris/ui-common';
import { FarrisLocalePipe } from './locale.pipe';
import { NgModule } from '@angular/core';
import { LocaleService, FARRIS_LOCAL_CUSTOM_DATA } from './locale.service';
var LocaleModule = /** @class */ (function () {
    function LocaleModule() {
    }
    /**
     * @param {?=} customData
     * @return {?}
     */
    LocaleModule.forRoot = /**
     * @param {?=} customData
     * @return {?}
     */
    function (customData) {
        return {
            ngModule: LocaleModule,
            providers: [
                { provide: FARRIS_LOCAL_CUSTOM_DATA, useValue: customData || {} },
                LocaleService
            ]
        };
    };
    LocaleModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        FarrisLocalePipe
                    ],
                    imports: [
                        FarrisCommonModule.forRoot()
                    ],
                    exports: [FarrisLocalePipe],
                    providers: [
                        LocaleService
                    ]
                },] }
    ];
    return LocaleModule;
}());
export { LocaleModule };
