// @qdp/query-framework/esm5/lib/component/qf-pagination/qf-pagination.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
var QueryFrameworkPaginationComponent = /** @class */ (function () {
    function QueryFrameworkPaginationComponent() {
        this.pagination = {
            pageIndex: 1,
            pageSize: 10,
            totalPage: 0
        };
        this.isDisablePreBtn = false;
        this.isDisableNextBtn = false;
        this.isEnable = true;
        this.totalCount = 21;
        this.pageSize = 10;
        this.pageIndex = 1;
        this.pageChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    QueryFrameworkPaginationComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initPageInfo();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    QueryFrameworkPaginationComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        for (var changeName in changes) {
            /** @type {?} */
            var newValue = changes[changeName].currentValue;
            if (newValue !== null && newValue !== void 0) {
                switch (changeName) {
                    case 'schemaInfo':
                        break;
                    case 'pageInfo':
                        if (newValue.pageSize === 0) {
                            this.isEnable = false;
                            return;
                        }
                        else {
                            this.isEnable = true;
                        }
                        this.pagination.pageIndex = this.pageIndex = newValue.pageIndex;
                        this.pagination.pageSize = this.pageSize = newValue.pageSize;
                        this.pagination.totalPage = newValue.pageCount;
                        this.totalCount = newValue.totalCount;
                        this.initPageInfo();
                        break;
                }
            }
        }
    };
    /**
     * @return {?}
     */
    QueryFrameworkPaginationComponent.prototype.initPageInfo = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var totalPageCount = 0;
        if (this.totalCount <= this.pageSize) {
            totalPageCount = 1;
        }
        else {
            totalPageCount = (this.totalCount % this.pageSize) > 0
                ? Math.ceil(this.totalCount / this.pageSize)
                : (this.totalCount / this.pageSize);
        }
        this.pagination = {
            pageIndex: this.pageIndex,
            totalPage: totalPageCount
        };
        this.isDisablePreBtn = this.pagination.pageIndex === 1 ? true : false;
        this.isDisableNextBtn = this.pagination.pageIndex === this.pagination.totalPage ? true : false;
        if (!this.isEnable) {
            this.isDisablePreBtn = this.isDisableNextBtn = !this.isEnable;
        }
    };
    /**
     * @param {?} type
     * @return {?}
     */
    QueryFrameworkPaginationComponent.prototype.handlePreBtnClick = /**
     * @param {?} type
     * @return {?}
     */
    function (type) {
        switch (type) {
            case '0':
                this.pagination.pageIndex = 1;
                break;
            case '1':
                this.pagination.pageIndex -= 1;
                break;
        }
        this.isDisablePreBtn = this.pagination.pageIndex === 1 ? true : false;
        this.isDisableNextBtn = this.pagination.pageIndex === this.pagination.totalPage ? true : false;
        this.pageChange.emit(this.pagination);
    };
    /**
     * @param {?} type
     * @return {?}
     */
    QueryFrameworkPaginationComponent.prototype.handleNextBtnClick = /**
     * @param {?} type
     * @return {?}
     */
    function (type) {
        switch (type) {
            case '0':
                this.pagination.pageIndex = this.pagination.totalPage;
                break;
            case '1':
                this.pagination.pageIndex += 1;
                break;
        }
        this.isDisablePreBtn = this.pagination.pageIndex === 1 ? true : false;
        this.isDisableNextBtn = this.pagination.pageIndex === this.pagination.totalPage ? true : false;
        this.pageChange.emit(this.pagination);
    };
    QueryFrameworkPaginationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'query-framework-pagination',
                    template: "<nav *ngIf=\"isEnable\" aria-label=\"Page navigation\">\r\n  <ul class=\"pagination p-0 pr-1\">\r\n    <li class=\"page-item\">\r\n      <span class=\"page-link border-0 w-100\">{{localizePageInfo?.totalPreInfo}}{{totalCount}}{{localizePageInfo?.totalEndInfo}}</span>\r\n    </li>\r\n    <li class=\"page-item ml-0\" [class.disabled]=\"isDisablePreBtn\">\r\n      <a href=\"javascript:void(0);\" (click)=\"handlePreBtnClick('0')\" tabindex=\"-1\" class=\"page-link border-0\">\r\n        <span class=\"f-icon f-icon-seek-w\" aria-label=\"\">\r\n        </span>\r\n      </a>\r\n    </li>\r\n    <li class=\"page-item\" [class.disabled]=\"isDisablePreBtn\">\r\n      <a href=\"javascript:void(0);\" (click)=\"handlePreBtnClick('1')\" tabindex=\"-1\" class=\"page-link border-0\">\r\n        <span class=\"f-icon f-icon-arrow-w\" aria-label=\"\">\r\n        </span>\r\n      </a>\r\n    </li>\r\n    <li class=\"page-item\">\r\n      <span class=\"page-link border-0 w-100\">{{pagination.pageIndex}}/{{pagination.totalPage?pagination.totalPage:0}}</span>\r\n    </li>\r\n    <li class=\"page-item\" [class.disabled]=\"isDisableNextBtn\">\r\n      <a href=\"javascript:void(0);\" (click)=\"handleNextBtnClick('1')\" tabindex=\"-1\" class=\"page-link border-0\">\r\n        <span class=\"f-icon f-icon-arrow-e\" aria-label=\"\">\r\n        </span>\r\n      </a>\r\n    </li>\r\n    <li class=\"page-item\" [class.disabled]=\"isDisableNextBtn\">\r\n      <a href=\"javascript:void(0);\" (click)=\"handleNextBtnClick('0')\" tabindex=\"-1\" class=\"page-link border-0\">\r\n        <span class=\"f-icon f-icon-seek-e\" aria-label=\"\">\r\n        </span>\r\n      </a>\r\n    </li>\r\n  </ul>\r\n</nav>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    QueryFrameworkPaginationComponent.ctorParameters = function () { return []; };
    QueryFrameworkPaginationComponent.propDecorators = {
        pageInfo: [{ type: Input }],
        localizePageInfo: [{ type: Input }],
        pageChange: [{ type: Output }]
    };
    return QueryFrameworkPaginationComponent;
}());
export { QueryFrameworkPaginationComponent };
if (false) {
    /** @type {?} */
    QueryFrameworkPaginationComponent.prototype.pagination;
    /** @type {?} */
    QueryFrameworkPaginationComponent.prototype.isDisablePreBtn;
    /** @type {?} */
    QueryFrameworkPaginationComponent.prototype.isDisableNextBtn;
    /** @type {?} */
    QueryFrameworkPaginationComponent.prototype.isEnable;
    /** @type {?} */
    QueryFrameworkPaginationComponent.prototype.pageInfo;
    /** @type {?} */
    QueryFrameworkPaginationComponent.prototype.localizePageInfo;
    /** @type {?} */
    QueryFrameworkPaginationComponent.prototype.totalCount;
    /** @type {?} */
    QueryFrameworkPaginationComponent.prototype.pageSize;
    /** @type {?} */
    QueryFrameworkPaginationComponent.prototype.pageIndex;
    /** @type {?} */
    QueryFrameworkPaginationComponent.prototype.pageChange;
}
