// @farris/ide-property-panel/esm5/lib/property-panel/controls/combolist.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output } from '@angular/core';
var CombolistPropertyComponent = /** @class */ (function () {
    function CombolistPropertyComponent() {
        this.valueChanged = new EventEmitter();
        /**
         * 下拉框的枚举项
         */
        this.selectOptions = [];
        /**
         * 是否可任意输入
         */
        this.nosearch = false;
        /**
         * 是否多选
         */
        this.multiSelect = false;
    }
    Object.defineProperty(CombolistPropertyComponent.prototype, "elementValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._elementValue;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._elementValue = value;
            // 多选控件使用数组格式的数据
            if (this.elementConfig.multiSelectDataType === 'array') {
                this._elementValue = this._elementValue ? this._elementValue.toString() : '';
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CombolistPropertyComponent.prototype, "elementConfig", {
        get: /**
         * @return {?}
         */
        function () {
            return this._elementConfig;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._elementConfig = value;
            if (!this.elementConfig) {
                return;
            }
            this.editType = this.elementConfig.propertyType;
            if (this.editType === 'boolean') {
                this.selectOptions = [{ key: true, value: 'true' }, { key: false, value: 'false' }];
                this.editType = 'select';
            }
            else {
                this.selectOptions = this.elementConfig.iterator || [];
            }
            switch (this.editType) {
                case 'select': {
                    // 不可编辑下拉框
                    this.editable = false;
                    break;
                }
                case 'editableSelect': {
                    // 可编辑下拉框
                    this.editable = true;
                    this.nosearch = true;
                    break;
                }
                case 'multiSelect': {
                    // 不可编辑的下拉多选框
                    this.editable = false;
                    this.multiSelect = true;
                    // 多选控件使用数组格式的数据
                    if (this.elementConfig.multiSelectDataType === 'array') {
                        this._elementValue = this._elementValue ? this._elementValue.toString() : '';
                    }
                    break;
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    CombolistPropertyComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} $event
     * @return {?}
     */
    CombolistPropertyComponent.prototype.changeValue = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var value = this._elementValue;
        // 多选控件返回数组格式的数据
        if (this.editType === 'multiSelect' && this.elementConfig.multiSelectDataType === 'array') {
            value = this._elementValue ? this._elementValue.split(',') : '';
        }
        this.valueChanged.next({ elementValue: value });
    };
    CombolistPropertyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-combolist-prop',
                    template: "\n      <!--\u4E0B\u62C9\u5217\u8868-->\n      <div class=\"farris-input-wrap\">\n        <farris-combo-list\n          [(ngModel)]=\"elementValue\"\n          [idField]=\"'key'\"\n          [textField]=\"'value'\"\n          [data]=\"selectOptions\"\n          [readonly]=\"elementConfig.readonly\"\n          [editable]=\"editable\"\n          [multiSelect]=\"multiSelect\"\n          [enableClear]=\"false\"\n          [enableCancelSelected]=\"false\"\n          [nosearch]=\"nosearch\"\n          (ngModelChange)=\"changeValue($event)\">\n        </farris-combo-list>\n      </div>\n    "
                }] }
    ];
    CombolistPropertyComponent.propDecorators = {
        valueChanged: [{ type: Output }]
    };
    return CombolistPropertyComponent;
}());
export { CombolistPropertyComponent };
if (false) {
    /** @type {?} */
    CombolistPropertyComponent.prototype.valueChanged;
    /**
     * @type {?}
     * @private
     */
    CombolistPropertyComponent.prototype._elementConfig;
    /**
     * @type {?}
     * @private
     */
    CombolistPropertyComponent.prototype._elementValue;
    /**
     * 编辑类型
     * @type {?}
     */
    CombolistPropertyComponent.prototype.editType;
    /**
     * 下拉框的枚举项
     * @type {?}
     */
    CombolistPropertyComponent.prototype.selectOptions;
    /**
     * 是否可编辑
     * @type {?}
     */
    CombolistPropertyComponent.prototype.editable;
    /**
     * 是否可任意输入
     * @type {?}
     */
    CombolistPropertyComponent.prototype.nosearch;
    /**
     * 是否多选
     * @type {?}
     */
    CombolistPropertyComponent.prototype.multiSelect;
}
