// @farris/ide-property-panel/esm5/lib/property-panel/controls/date.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, EventEmitter, Output } from '@angular/core';
var DatePropertyComponent = /** @class */ (function () {
    function DatePropertyComponent(cd) {
        this.cd = cd;
        this.showTime = false;
        this.dateFormat = 'yyyy-MM-dd';
        this.valueChanged = new EventEmitter();
    }
    Object.defineProperty(DatePropertyComponent.prototype, "elementConfig", {
        get: /**
         * @return {?}
         */
        function () {
            return this._elementConfig;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._elementConfig = value;
            if (!this.elementConfig) {
                return;
            }
            this.maxDate = value.max;
            this.minDate = value.min;
            this.showTime = this.elementConfig.propertyType === 'datetime';
            this.dateFormat = this.elementConfig.propertyType === 'datetime' ? 'yyyy-MM-dd HH:mm:ss' : 'yyyy-MM-dd';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePropertyComponent.prototype, "elementValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._elementValue;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._elementValue = value;
        },
        enumerable: true,
        configurable: true
    });
    // editType: string;
    // editType: string;
    /**
     * @return {?}
     */
    DatePropertyComponent.prototype.ngOnInit = 
    // editType: string;
    /**
     * @return {?}
     */
    function () {
        // this.editType = this.elementConfig && this.elementConfig.propertyType;
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    DatePropertyComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatePropertyComponent.prototype.changeValue = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var value = $event.formatted;
        this.valueChanged.next({ $event: $event, elementValue: value });
    };
    DatePropertyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-date-prop',
                    template: "\n        <farris-datepicker [(ngModel)]=\"elementValue\" [editable]=\"false\" [showTime]=\"showTime\"\n          [dateFormat]=\"dateFormat\" [readonly]=\"elementConfig.readonly\"\n          (valueChange)=\"changeValue($event)\" [returnType]=\"'string'\" [maxDate]=\"maxDate\"\n          [minDate]=\"minDate\">\n        </farris-datepicker>\n\n    "
                }] }
    ];
    /** @nocollapse */
    DatePropertyComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    DatePropertyComponent.propDecorators = {
        valueChanged: [{ type: Output }]
    };
    return DatePropertyComponent;
}());
export { DatePropertyComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DatePropertyComponent.prototype._elementConfig;
    /**
     * @type {?}
     * @private
     */
    DatePropertyComponent.prototype._elementValue;
    /** @type {?} */
    DatePropertyComponent.prototype.maxDate;
    /** @type {?} */
    DatePropertyComponent.prototype.minDate;
    /** @type {?} */
    DatePropertyComponent.prototype.showTime;
    /** @type {?} */
    DatePropertyComponent.prototype.dateFormat;
    /** @type {?} */
    DatePropertyComponent.prototype.valueChanged;
    /**
     * @type {?}
     * @private
     */
    DatePropertyComponent.prototype.cd;
}
