// @farris/ide-property-panel/esm5/lib/property-panel/controls/number.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output } from '@angular/core';
var NumberPropertyComponent = /** @class */ (function () {
    function NumberPropertyComponent() {
        this.valueChanged = new EventEmitter();
        /**
         * 数字控件精度
         */
        this.precision = 2;
    }
    Object.defineProperty(NumberPropertyComponent.prototype, "elementConfig", {
        get: /**
         * @return {?}
         */
        function () {
            return this._elementConfig;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._elementConfig = value;
            this.precision = this._elementConfig.decimals === undefined ? 2 : this._elementConfig.decimals;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NumberPropertyComponent.prototype, "elementValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._elementValue;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._elementValue = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NumberPropertyComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} $event
     * @return {?}
     */
    NumberPropertyComponent.prototype.changeValue = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.valueChanged.next({ $event: $event, elementValue: this.elementValue });
    };
    NumberPropertyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-number-prop',
                    template: "\n    <div class=\"farris-input-wrap\">\n        <farris-number-spinner [(ngModel)]=\"elementValue\" [readonly]=\"elementConfig.readonly\" [precision]=\"precision\"\n            [autoDecimal]=\"true\" [min]=\"elementConfig.min\" [max]=\"elementConfig.max\" [canNull]=\"true\"\n            [bigNumber]=\"elementConfig.isBigNumber\" (ngModelChange)=\"changeValue($event)\">\n        </farris-number-spinner>\n    </div>\n    "
                }] }
    ];
    NumberPropertyComponent.propDecorators = {
        valueChanged: [{ type: Output }]
    };
    return NumberPropertyComponent;
}());
export { NumberPropertyComponent };
if (false) {
    /** @type {?} */
    NumberPropertyComponent.prototype.valueChanged;
    /**
     * 数字控件精度
     * @type {?}
     */
    NumberPropertyComponent.prototype.precision;
    /**
     * @type {?}
     * @private
     */
    NumberPropertyComponent.prototype._elementConfig;
    /**
     * @type {?}
     * @private
     */
    NumberPropertyComponent.prototype._elementValue;
}
