// @farris/ide-property-panel/esm5/lib/property-panel/controls/textbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output } from '@angular/core';
var TextboxPropertyComponent = /** @class */ (function () {
    function TextboxPropertyComponent() {
        this.valueChanged = new EventEmitter();
    }
    Object.defineProperty(TextboxPropertyComponent.prototype, "elementConfig", {
        get: /**
         * @return {?}
         */
        function () {
            return this._elementConfig;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._elementConfig = value;
            if (this.elementConfig.readonly) {
                this.convertModalShowValue(this._elementValue);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextboxPropertyComponent.prototype, "elementValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._elementValue;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._elementValue = value;
            if (this.elementConfig.readonly) {
                this.convertModalShowValue(this._elementValue);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TextboxPropertyComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    TextboxPropertyComponent.prototype.changeValue = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        var _this = this;
        // 文本控件限制输入字符
        if (this.elementValue && !this.elementConfig.readonly &&
            this.elementConfig.notAllowedChars && this.elementConfig.notAllowedChars.length) {
            this.elementConfig.notAllowedChars.forEach((/**
             * @param {?} c
             * @return {?}
             */
            function (c) {
                if (typeof (c) === 'object') {
                    _this.elementValue = _this.elementValue.replace(c, '');
                }
                else if (typeof (c) === 'string') {
                    _this.elementValue = _this.elementValue.replace(new RegExp(c, 'g'), '');
                }
            }));
        }
        this.valueChanged.next({ $event: $event, elementValue: this.elementValue });
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TextboxPropertyComponent.prototype.restValueAfterValidRestrict = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.elementValue = value || '';
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    TextboxPropertyComponent.prototype.convertModalShowValue = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this.elementConfig.converter) {
            this.elementShowValue = this.elementConfig.converter.convertTo(value);
            return;
        }
        if (value && value instanceof Object) {
            this.elementShowValue = JSON.stringify(value);
            return;
        }
        this.elementShowValue = value;
    };
    TextboxPropertyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-textbox-prop',
                    template: "\n    <div class=\"farris-input-wrap\">\n        <input type=\"input\" class=\"form-control form-control-sm\" [(ngModel)]=\"elementValue\" (blur)=\"changeValue($event)\"\n            *ngIf=\"!elementConfig.readonly\">\n        <input type=\"input\" class=\"form-control form-control-sm\" [(ngModel)]=\"elementShowValue\" [readonly]=\"'readonly'\"\n            *ngIf=\"elementConfig.readonly\">\n    </div>\n    "
                }] }
    ];
    TextboxPropertyComponent.propDecorators = {
        valueChanged: [{ type: Output }]
    };
    return TextboxPropertyComponent;
}());
export { TextboxPropertyComponent };
if (false) {
    /** @type {?} */
    TextboxPropertyComponent.prototype.valueChanged;
    /**
     * @type {?}
     * @private
     */
    TextboxPropertyComponent.prototype._elementConfig;
    /**
     * @type {?}
     * @private
     */
    TextboxPropertyComponent.prototype._elementValue;
    /**
     * 输入框中显示的值
     * @type {?}
     */
    TextboxPropertyComponent.prototype.elementShowValue;
}
