// @farris/ide-property-panel/esm5/lib/property-panel/controls/unity.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { PropertyEntity } from '../entity/property-entity';
var UniformPropertyComponent = /** @class */ (function () {
    function UniformPropertyComponent() {
        this.valueChanged = new EventEmitter();
    }
    /**
     * @return {?}
     */
    UniformPropertyComponent.prototype.initImportedOriginalData = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var editorParams = this.elementConfig.editorParams;
        /** @type {?} */
        var constType = editorParams.constType;
        /** @type {?} */
        var valueType = this.getValueType();
        this.importedOriginalData = {
            controlName: editorParams.controlName,
            propertyName: this.elementConfig.propertyID,
            propertyType: constType,
            propertyValue: {
                type: valueType,
                value: this.elementValue
            },
            editorOptions: editorParams.editorOptions
        };
    };
    /**
     * 通过当前配置信息和值信息推断当前值的类型
     * @returns 值类型
     */
    /**
     * 通过当前配置信息和值信息推断当前值的类型
     * @return {?} 值类型
     */
    UniformPropertyComponent.prototype.getValueType = /**
     * 通过当前配置信息和值信息推断当前值的类型
     * @return {?} 值类型
     */
    function () {
        /** @type {?} */
        var propertyValue = this.elementValue;
        /** @type {?} */
        var editorParams = this.elementConfig.editorParams;
        /** @type {?} */
        var constType = editorParams.constType;
        /** @type {?} */
        var types = editorParams.editorOptions.types;
        // 当前值为对象时
        if (propertyValue && typeof propertyValue === "object") {
            // 判断是否为变量类型
            if (propertyValue.hasOwnProperty("type") && (typeof propertyValue.type === "string" && propertyValue.type.toLowerCase() === "variable")) {
                if (types.includes("variable"))
                    return "variable";
            }
        }
        // 当前值为字符串时
        if (typeof propertyValue === "string") {
            // 判断是否为自定义类型（通过是否含有“viewModel”子串硬推断）
            if (types.includes("custom") && (propertyValue.includes("viewModel") || !types.includes("const") || constType !== "string")) {
                return "custom";
            }
            // 判断是否为字符串常量
            if (types.includes("const") && constType === "string") {
                return "const";
            }
        }
        // 当前值为数字时
        if (typeof propertyValue === "number") {
            if (types.includes("const") && constType === "number") {
                return "const";
            }
        }
        // 默认为const类型（为枚举类型兜底）
        if (types.includes("const")) {
            return "const";
        }
        // 如果都找不到则任意返回一个值
        return types[0];
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    UniformPropertyComponent.prototype.handleValueChanged = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var newValue = $event.propertyValue.value;
        if ($event.propertyValue.type === "variable" && typeof newValue === "object") {
            newValue.type = "Variable";
            if (newValue.name) {
                delete newValue.name;
            }
            if ($event["isNewVariable"]) {
                newValue["isNewVariable"] = true;
            }
        }
        this.valueChanged.next({ $event: $event, elementValue: newValue });
    };
    /**
     * @return {?}
     */
    UniformPropertyComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initImportedOriginalData();
    };
    UniformPropertyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-unity-prop',
                    template: "\n      <div class=\"farris-input-wrap\">\n        <farris-propertyEditor\n          [importedOriginalData]=\"importedOriginalData\"\n          (propertyEditorValueChanged)=\"handleValueChanged($event)\">\n        </farris-propertyEditor>\n      </div>\n    "
                }] }
    ];
    UniformPropertyComponent.propDecorators = {
        elementConfig: [{ type: Input }],
        elementValue: [{ type: Input }],
        valueChanged: [{ type: Output }]
    };
    return UniformPropertyComponent;
}());
export { UniformPropertyComponent };
if (false) {
    /** @type {?} */
    UniformPropertyComponent.prototype.elementConfig;
    /** @type {?} */
    UniformPropertyComponent.prototype.elementValue;
    /** @type {?} */
    UniformPropertyComponent.prototype.valueChanged;
    /** @type {?} */
    UniformPropertyComponent.prototype.importedOriginalData;
}
