// @gsp-cmp/ubcc-param-data-ui/esm2015/lib/controls/dimension-control/dimension-lookup-control/dimension-lookup-control.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/controls/dimension-control/dimension-lookup-control/dimension-lookup-control.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { GSPMetadataRTService } from '@gsp-lcm/metadatart-selector';
import { ServerSideToken } from '@farris/ui-lookup';
import { LookUpHttpService } from '../../../service/lookup-http-service';
import { LookUpBindingFields } from '../../../common/look-up-binding-fields';
import { UbccEventBusService } from '../../../service/event-bus-service';
/** @type {?} */
const BCC_HELP_URL = `/api/runtime/bcc/v1.0/param/help/{helpId}`;
export class DimensionLookupControlComponent {
    /**
     * @param {?} metadataRTService
     * @param {?} ebsService
     */
    constructor(metadataRTService, ebsService) {
        this.metadataRTService = metadataRTService;
        this.ebsService = ebsService;
        this.isSingleSelect = true;
        this.isHelpInfoLoad = false;
        this.helpUrl = BCC_HELP_URL;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.dimensionItem = this.dimensionItem || {};
        this.bindingFields = this.bindingFields || new LookUpBindingFields();
        this.helpConfigInit();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        this.helpConfigInit();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    selectedHelpHandler($event) {
        if (!$event) {
            return;
        }
        this.dimensionItem.value = '';
        this.dimensionItem.value = $event[this.bindingFields.idField];
        this.ebsService.ebs.emit(this.dimensionItem);
    }
    /**
     * @private
     * @return {?}
     */
    helpConfigInit() {
        if (this.isHelpInfoLoad || !this.dimensionItem || !this.dimensionItem.dataType) {
            return;
        }
        /** @type {?} */
        let helpType = (/** @type {?} */ (this.dimensionItem.dataType));
        this.loadDataFromMetadata(helpType);
    }
    /**
     * @private
     * @param {?} helpType
     * @return {?}
     */
    loadDataFromMetadata(helpType) {
        if (!helpType.helpId) {
            return;
        }
        this.metadataRTService.GetMetadataRT(helpType.helpId).subscribe((/**
         * @param {?} metadataDto
         * @return {?}
         */
        (metadataDto) => {
            /** @type {?} */
            let helpMetadata = JSON.parse(metadataDto.content);
            this.setHelpInfo(helpMetadata);
            this.isHelpInfoLoad = true;
        }));
    }
    /**
     * @param {?} helpMetadata
     * @return {?}
     */
    setHelpInfo(helpMetadata) {
        this.helpUrl = BCC_HELP_URL.replace('{helpId}', helpMetadata.id);
        this.bindingFields.idField = helpMetadata.idField;
        this.bindingFields.textField = helpMetadata.textField;
        this.bindingFields.valueField = helpMetadata.valueField;
        this.bindingFields.displayType = helpMetadata.displayType;
    }
}
DimensionLookupControlComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-dimension-lookup-control',
                template: "<div class=\"farris-form-controls-inline farris-group-wrap col-12\">\n  <div class=\"form-group farris-form-group\">\n    <label class=\"col-form-label col-4\">\n      <span class=\"farris-label-text\" style=\"float:right\">{{dimensionItem.name}}</span>\n    </label>\n    <div class=\"farris-input-wrap col-6\">\n      <farris-lookup-grid [editable]=\"false\" (selectedData)=\"selectedHelpHandler($event)\" [uri]=\"helpUrl\" [idField]=\"bindingFields.idField\"\n        [textField]=\"bindingFields.textField\" [valueField]=\"bindingFields.valueField\" [enableClear]=\"false\" \n        [(ngModel)]=\"dimensionItem.valueText\" name=\"{{dimensionItem.code}}\"\n        [displayType]=\"bindingFields.displayType\" [pagination]=\"true\" [singleSelect]=\"isSingleSelect\" required>\n      </farris-lookup-grid>\n    </div>\n  </div>\n</div>",
                providers: [{ provide: ServerSideToken, useClass: LookUpHttpService }, GSPMetadataRTService],
                styles: [""]
            }] }
];
/** @nocollapse */
DimensionLookupControlComponent.ctorParameters = () => [
    { type: GSPMetadataRTService },
    { type: UbccEventBusService }
];
DimensionLookupControlComponent.propDecorators = {
    dimensionItem: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DimensionLookupControlComponent.prototype.dimensionItem;
    /** @type {?} */
    DimensionLookupControlComponent.prototype.isSingleSelect;
    /** @type {?} */
    DimensionLookupControlComponent.prototype.isHelpInfoLoad;
    /** @type {?} */
    DimensionLookupControlComponent.prototype.bindingFields;
    /** @type {?} */
    DimensionLookupControlComponent.prototype.helpUrl;
    /**
     * @type {?}
     * @private
     */
    DimensionLookupControlComponent.prototype.metadataRTService;
    /**
     * @type {?}
     * @private
     */
    DimensionLookupControlComponent.prototype.ebsService;
}
