// @edp-aif/runtime-api/esm5/lib/entity/rpc/RpcCreationRuleExecutorResult.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { RpcCreationMappingResult } from './RpcCreationMappingResult';
var RpcCreationRuleExecutorResult = /** @class */ (function (_super) {
    tslib_1.__extends(RpcCreationRuleExecutorResult, _super);
    function RpcCreationRuleExecutorResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.mappingResults = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RpcCreationRuleExecutorResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.creationRuleId = null;
        this.creationRuleInstanceId = null;
        this.bizFlowchartId = null;
        this.mappingResults = new Array();
    };
    /**
     * @return {?}
     */
    RpcCreationRuleExecutorResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingResults", this.mappingResults);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RpcCreationRuleExecutorResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.creationRuleInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.mappingResults = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingResults", RpcCreationMappingResult);
    };
    return RpcCreationRuleExecutorResult;
}(JsonConvertableImpl));
export { RpcCreationRuleExecutorResult };
if (false) {
    /** @type {?} */
    RpcCreationRuleExecutorResult.prototype.creationRuleId;
    /** @type {?} */
    RpcCreationRuleExecutorResult.prototype.creationRuleInstanceId;
    /** @type {?} */
    RpcCreationRuleExecutorResult.prototype.bizFlowchartId;
    /** @type {?} */
    RpcCreationRuleExecutorResult.prototype.mappingResults;
}
