// @edp-aif/runtime-api/esm5/lib/entity/rpc/RpcCreationRuleExecutorArgs.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
var RpcCreationRuleExecutorArgs = /** @class */ (function (_super) {
    tslib_1.__extends(RpcCreationRuleExecutorArgs, _super);
    function RpcCreationRuleExecutorArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.sourceVoEntities = new Array();
        _this.targetVoProxyEntities = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RpcCreationRuleExecutorArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.sourceVoEntities = new Array();
        this.creationRuleId = null;
        this.bizFlowchartId = null;
        this.sourceVoId = null;
        this.targetVoProxyId = null;
        this.targetVoProxyEntities = new Array();
    };
    /**
     * @return {?}
     */
    RpcCreationRuleExecutorArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntities", BpmnModelJsonHelper.StringListToJson(this.sourceVoEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyId", this.targetVoProxyId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyEntities", BpmnModelJsonHelper.StringListToJson(this.targetVoProxyEntities));
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RpcCreationRuleExecutorArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.sourceVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.sourceVoEntities = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoEntities"));
        this.targetVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyId");
        this.targetVoProxyEntities = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyEntities"));
    };
    return RpcCreationRuleExecutorArgs;
}(JsonConvertableImpl));
export { RpcCreationRuleExecutorArgs };
if (false) {
    /** @type {?} */
    RpcCreationRuleExecutorArgs.prototype.creationRuleId;
    /** @type {?} */
    RpcCreationRuleExecutorArgs.prototype.sourceVoId;
    /** @type {?} */
    RpcCreationRuleExecutorArgs.prototype.bizFlowchartId;
    /** @type {?} */
    RpcCreationRuleExecutorArgs.prototype.sourceVoEntities;
    /** @type {?} */
    RpcCreationRuleExecutorArgs.prototype.targetVoProxyId;
    /** @type {?} */
    RpcCreationRuleExecutorArgs.prototype.targetVoProxyEntities;
}
