// @edp-aif/runtime-api/esm5/lib/entity/rpc/RpcFireBizEventArgs.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { CreationRuleExecutorAssociations } from '../CreationRuleExecutorAssociations';
var RpcFireBizEventArgs = /** @class */ (function (_super) {
    tslib_1.__extends(RpcFireBizEventArgs, _super);
    function RpcFireBizEventArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.voChangeDetails = null;
        _this.voEntityDatas = null;
        _this.voEntityDataIds = new Array();
        _this.associations = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RpcFireBizEventArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.voId = null;
        this.voChangeDetails = null;
        this.voEntityDatas = null;
        this.voEntityDataIds = new Array();
        this.bizEventId = null;
        this.voFuncSessionId = null;
        this.bizFlowchartId = null;
        this.upperBeFuncSessionId = null;
        this.associations = new Array();
    };
    /**
     * @return {?}
     */
    RpcFireBizEventArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voChangeDetails", this.voChangeDetails);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voEntityDatas", this.voEntityDatas);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voEntityDataIds", BpmnModelJsonHelper.StringListToJsonArray(this.voEntityDataIds));
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventId", this.bizEventId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voFuncSessionId", this.voFuncSessionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upperBeFuncSessionId", this.upperBeFuncSessionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RpcFireBizEventArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
        this.voChangeDetails = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voChangeDetails");
        this.voEntityDatas = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voEntityDatas");
        this.voEntityDataIds = BpmnModelJsonHelper.StringListFromJsonArray(BpmnModelJsonHelper.GetAttributeValue(obj, "voEntityDataIds"));
        this.bizEventId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventId");
        this.voFuncSessionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voFuncSessionId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.upperBeFuncSessionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upperBeFuncSessionId");
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", CreationRuleExecutorAssociations);
    };
    return RpcFireBizEventArgs;
}(JsonConvertableImpl));
export { RpcFireBizEventArgs };
if (false) {
    /** @type {?} */
    RpcFireBizEventArgs.prototype.voId;
    /** @type {?} */
    RpcFireBizEventArgs.prototype.voChangeDetails;
    /** @type {?} */
    RpcFireBizEventArgs.prototype.voEntityDatas;
    /** @type {?} */
    RpcFireBizEventArgs.prototype.voEntityDataIds;
    /** @type {?} */
    RpcFireBizEventArgs.prototype.bizEventId;
    /** @type {?} */
    RpcFireBizEventArgs.prototype.voFuncSessionId;
    /** @type {?} */
    RpcFireBizEventArgs.prototype.bizFlowchartId;
    /** @type {?} */
    RpcFireBizEventArgs.prototype.upperBeFuncSessionId;
    /** @type {?} */
    RpcFireBizEventArgs.prototype.associations;
}
