// @edp-aif/runtime-api/esm5/lib/entity/rpc/RpcWritebackRuleExecutorArgs.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { MappingAssociation } from '../MappingAssociation';
var RpcWritebackRuleExecutorArgs = /** @class */ (function (_super) {
    tslib_1.__extends(RpcWritebackRuleExecutorArgs, _super);
    function RpcWritebackRuleExecutorArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.sourceVoEntities = new Array();
        _this.associations = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RpcWritebackRuleExecutorArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.writebackRuleId = null;
        this.sourceVoId = null;
        this.targetVoFuncSessionId = null;
        this.sourceVoEntities = new Array();
        this.associations = new Array();
    };
    /**
     * @return {?}
     */
    RpcWritebackRuleExecutorArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "writebackRuleId", this.writebackRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoFuncSessionId", this.targetVoFuncSessionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntities", BpmnModelJsonHelper.StringListToJson(this.sourceVoEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RpcWritebackRuleExecutorArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.writebackRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackRuleId");
        this.sourceVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
        this.targetVoFuncSessionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoFuncSessionId");
        this.sourceVoEntities = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoEntities"));
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
    };
    return RpcWritebackRuleExecutorArgs;
}(JsonConvertableImpl));
export { RpcWritebackRuleExecutorArgs };
if (false) {
    /** @type {?} */
    RpcWritebackRuleExecutorArgs.prototype.writebackRuleId;
    /** @type {?} */
    RpcWritebackRuleExecutorArgs.prototype.sourceVoId;
    /** @type {?} */
    RpcWritebackRuleExecutorArgs.prototype.sourceVoEntities;
    /** @type {?} */
    RpcWritebackRuleExecutorArgs.prototype.associations;
    /** @type {?} */
    RpcWritebackRuleExecutorArgs.prototype.targetVoFuncSessionId;
}
