// @farris/ui-forms/fesm2015/farris-ui-forms-form-header.js
import { CommonModule } from '@angular/common';
import { Component, Input, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class FormHeaderComponent {
    constructor() { }
    /**
     * @return {?}
     */
    ngOnInit() { }
}
FormHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'form-header',
                template: "<div class=\"form-header-bar\">\r\n    <!-- <span>icon</span> -->\r\n    <!-- <h5>\u5355\u636E\u7F16\u53F7:<span></span><span class=\"state success\">\u5BA1\u6279\u901A\u8FC7</span></h5> -->\r\n    <div class=\"btn-bar\">\r\n        <span class=\"k-icon k-i-arrow-60-left\" (click)='prev()'></span>\r\n        <span class=\"k-icon k-i-arrow-60-right\" (click)='next()'></span>\r\n    </div>\r\n</div>",
                styles: [".form-header-bar{display:inline-block;margin-left:30px;vertical-align:top}.form-header-bar h5{font-size:18px;color:#000;line-height:28px;display:inline-block;text-align:left;margin:0 0 0 22px}.form-header-bar span.state{font-size:12px;color:rgba(255,255,255,.85);text-align:left;padding:2px 5px;border-radius:2px;vertical-align:top;margin-left:7px}.form-header-bar span.state.success{background-color:#299951e6}.form-header-bar .btn-bar{display:inline-block;vertical-align:bottom;margin-left:7px;line-height:28px}.form-header-bar .btn-bar .k-icon{font-size:16px;line-height:19px;height:20px;width:20px;padding:0;margin-left:7px;text-align:center;color:#bebebe;border:1px solid #dcdcdc;border-radius:2px;vertical-align:text-top;cursor:pointer}.form-header-bar .btn-bar .k-icon.diable{color:#d0d0d0;background-color:#f6f6f6;cursor:not-allowed}"]
            }] }
];
/** @nocollapse */
FormHeaderComponent.ctorParameters = () => [];
FormHeaderComponent.propDecorators = {
    prev: [{ type: Input }],
    next: [{ type: Input }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class FormHeaderModule {
}
FormHeaderModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    FormHeaderComponent
                ],
                exports: [
                    FormHeaderComponent
                ]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { FormHeaderModule, FormHeaderComponent };
