// @gsp-sys/rac-common/esm5/lib/common/util/rac-util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpHeaders } from '@angular/common/http';
import { SessionService } from '@ecp-caf/caf-common';
import { BizDataDTO } from '../entity/auresult.entity';
var RacCommonUtil = /** @class */ (function () {
    function RacCommonUtil(sessionSv) {
        this.sessionSv = sessionSv;
        /**
         * 维度1的值
         */
        this.assid1 = '';
        /**
         * 维度2的值
         */
        this.assid2 = '';
        this.AUDATACONTROL = 'AuDataControl';
        this.expandIcon = 'f-icon f-icon-file-folder-open text-warning';
        this.collapseIcon = 'f-icon f-icon-file-folder-close text-warning';
        this.leafIcon = 'f-icon f-icon-file-txt text-warning';
    }
    /**
     * 构造Header
     */
    /**
     * 构造Header
     * @return {?}
     */
    RacCommonUtil.prototype.getHeader = /**
     * 构造Header
     * @return {?}
     */
    function () {
        /** @type {?} */
        var sessionId = this.sessionSv.getCurrentSeesionId();
        /** @type {?} */
        var headers = new HttpHeaders().set('SessionId', sessionId);
        return { headers: headers };
    };
    /**
     * 获取上下文中的用户
     */
    /**
     * 获取上下文中的用户
     * @return {?}
     */
    RacCommonUtil.prototype.getCurrentUserId = /**
     * 获取上下文中的用户
     * @return {?}
     */
    function () {
        // 测试用写死
        return 'admin';
    };
    /** Guid生成 */
    /**
     * Guid生成
     * @return {?}
     */
    RacCommonUtil.prototype.getGuid36 = /**
     * Guid生成
     * @return {?}
     */
    function () {
        /**
         * @return {?}
         */
        function S4() {
            // tslint:disable-next-line:no-bitwise
            return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
        }
        return (S4() + S4() + '-' + S4() + '-' + S4() + '-' + S4() + '-' + S4() + S4() + S4());
    };
    /**
     * 权限分配结果路径
     */
    /**
     * 权限分配结果路径
     * @return {?}
     */
    RacCommonUtil.prototype.getResultPath = /**
     * 权限分配结果路径
     * @return {?}
     */
    function () {
        /** @type {?} */
        var path = "/api/runtime/sys/v1.0/assignresults";
        return path;
    };
    /**
     * 权限对象路径
     */
    /**
     * 权限对象路径
     * @return {?}
     */
    RacCommonUtil.prototype.getAuthorizationPath = /**
     * 权限对象路径
     * @return {?}
     */
    function () {
        /** @type {?} */
        var path = "/api/runtime/sys/v1.0/authorizations";
        return path;
    };
    /**
     * 权限字段路径
     */
    /**
     * 权限字段路径
     * @return {?}
     */
    RacCommonUtil.prototype.getAuthfieldPath = /**
     * 权限字段路径
     * @return {?}
     */
    function () {
        /** @type {?} */
        var path = "/api/runtime/sys/v1.0/authfields";
        return path;
    };
    /**
     * 重授权对象路径
     */
    /**
     * 重授权对象路径
     * @return {?}
     */
    RacCommonUtil.prototype.getReAssignPath = /**
     * 重授权对象路径
     * @return {?}
     */
    function () {
        /** @type {?} */
        var path = "/api/runtime/sys/v1.0/reassigns";
        return path;
    };
    /**
     * 获取新增的内容
     * @param newarr: 最新的内容
     * @param oldarr：历史内容
     */
    /**
     * 获取新增的内容
     * @param {?} newarr
     * @param {?} oldarr
     * @return {?}
     */
    RacCommonUtil.prototype.getAddIds = /**
     * 获取新增的内容
     * @param {?} newarr
     * @param {?} oldarr
     * @return {?}
     */
    function (newarr, oldarr) {
        /** @type {?} */
        var addIds = [];
        newarr.forEach((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            if (!oldarr.includes(x)) {
                addIds.push(x);
            }
        }));
        return addIds;
    };
    /**
     * 获取删除的内容
     * @param newarr: 最新的内容
     * @param oldarr：历史内容
     */
    /**
     * 获取删除的内容
     * @param {?} newarr
     * @param {?} oldarr
     * @return {?}
     */
    RacCommonUtil.prototype.getRemoveIds = /**
     * 获取删除的内容
     * @param {?} newarr
     * @param {?} oldarr
     * @return {?}
     */
    function (newarr, oldarr) {
        return this.getAddIds(oldarr, newarr);
    };
    /**
     * @param {?} arrs
     * @return {?}
     */
    RacCommonUtil.prototype.convertToBizDataDTO = /**
     * @param {?} arrs
     * @return {?}
     */
    function (arrs) {
        return arrs.map((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            /** @type {?} */
            var dataDto = new BizDataDTO();
            dataDto.data = x;
            return dataDto;
        }));
    };
    /**
     * 有效的字符串
     */
    /**
     * 有效的字符串
     * @param {?} value
     * @return {?}
     */
    RacCommonUtil.prototype.isNullOrEmptyString = /**
     * 有效的字符串
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return this.isBlank(value) || value.toString().trim().length === 0;
    };
    /**
     * 不为空
     */
    /**
     * 不为空
     * @param {?} value
     * @return {?}
     */
    RacCommonUtil.prototype.isBlank = /**
     * 不为空
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value === null || value === undefined;
    };
    /**
     * 字符串不为空
     */
    /**
     * 字符串不为空
     * @param {?} value
     * @return {?}
     */
    RacCommonUtil.prototype.isNotNullOrEmptyString = /**
     * 字符串不为空
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return !(this.isNullOrEmptyString(value));
    };
    /** 找到树形Parent分级方式的加载的第一层 */
    /**
     * 找到树形Parent分级方式的加载的第一层
     * @param {?} datas
     * @param {?} key
     * @param {?} parentKey
     * @return {?}
     */
    RacCommonUtil.prototype.findLayer1ByParentId = /**
     * 找到树形Parent分级方式的加载的第一层
     * @param {?} datas
     * @param {?} key
     * @param {?} parentKey
     * @return {?}
     */
    function (datas, key, parentKey) {
        /** @type {?} */
        var rtn = [];
        datas.forEach((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var findParent = datas.find((/**
             * @param {?} parent
             * @return {?}
             */
            function (parent) { return parent[key] === data[parentKey]; }));
            if (!findParent && data.type !== undefined && data.type !== null && data.type.toString() === '99') {
                data.parentId = undefined;
                rtn.push(data);
            }
        }));
        return rtn;
    };
    /** 类型是否为导航 */
    /**
     * 类型是否为导航
     * @param {?} type
     * @return {?}
     */
    RacCommonUtil.prototype.isNav = /**
     * 类型是否为导航
     * @param {?} type
     * @return {?}
     */
    function (type) {
        return type === '99' || type === '0';
    };
    /**
     * @param {?} farrisTt
     * @return {?}
     */
    RacCommonUtil.prototype.setTreeIcon = /**
     * @param {?} farrisTt
     * @return {?}
     */
    function (farrisTt) {
        farrisTt.expandIcon = this.expandIcon;
        farrisTt.collapseIcon = this.collapseIcon;
        farrisTt.leafIcon = this.leafIcon;
        farrisTt.detectChanges();
    };
    RacCommonUtil.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    RacCommonUtil.ctorParameters = function () { return [
        { type: SessionService }
    ]; };
    return RacCommonUtil;
}());
export { RacCommonUtil };
if (false) {
    /** @type {?} */
    RacCommonUtil.prototype.sessionId;
    /** @type {?} */
    RacCommonUtil.prototype.ownerId;
    /** @type {?} */
    RacCommonUtil.prototype.ownerCode;
    /** @type {?} */
    RacCommonUtil.prototype.ownerName;
    /**
     * 权限所有者类型 岗位1，用户2
     * @type {?}
     */
    RacCommonUtil.prototype.ownerType;
    /**
     * 岗位类型 组织岗0，通用岗1，用户2 TODO 命令不支持传参数，暂时这么处理
     * @type {?}
     */
    RacCommonUtil.prototype.posType;
    /** @type {?} */
    RacCommonUtil.prototype.packageId;
    /** @type {?} */
    RacCommonUtil.prototype.year;
    /**
     * 维度1的值
     * @type {?}
     */
    RacCommonUtil.prototype.assid1;
    /**
     * 维度2的值
     * @type {?}
     */
    RacCommonUtil.prototype.assid2;
    /** @type {?} */
    RacCommonUtil.prototype.funcId;
    /** @type {?} */
    RacCommonUtil.prototype.AUDATACONTROL;
    /** @type {?} */
    RacCommonUtil.prototype.expandIcon;
    /** @type {?} */
    RacCommonUtil.prototype.collapseIcon;
    /** @type {?} */
    RacCommonUtil.prototype.leafIcon;
    /**
     * @type {?}
     * @private
     */
    RacCommonUtil.prototype.sessionSv;
}
