// @farris/extend-header-footer/esm2015/lib/models/page-header.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/models/page-header.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const FEPageHeaderDefault = {
    mainTitle: '',
    // 主标题
    subTitle: '',
    // 副标题
    status: '',
    // 状态
    content: '',
    // 中间区域
    extend: '',
    // 扩展区域内容
    showIcon: true,
    //是否显示图标
    iconName: 'f-icon-page-title-record',
    //图标的名称
    iconCls: 'f-text-orna-bill',
    //图标的样式比如颜色、背景
    pgShow: false,
    //是否显示分页
    pgPreDisabled: true,
    //向前翻页是否禁用
    pgNextDisabled: false,
    //向后翻页是否禁用
    toolbarCls: 'col-6',
    //工具栏样式
    toolbarDatas: [],
    //工具栏数据
    toolbarBtnSize: "",
    //工具栏按钮尺寸
    toolbarPopDirection: "",
    //工具栏按钮弹出方向
    showViewChange: false,
    //是否显示视图方案
    viewDatas: [],
    //视图切换数据
    viewType: 'tile',
    // 视图切换类型
    viewDefaultType: '',
    // 默认视图状态
    viewGroupId: 'default',
    //视图切换groupId
    toolbarAllDatas: [],
    // 所有按钮相关数据
    toolbarConfig: null,
    //工具栏配置
    toolbarKey: "" //工具栏获取数据的依据
};
