// @gsp-lcm/bo-creator/esm5/lib/bo-creator-package.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bo-creator-package.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BOCreatorComponent } from './views/bo-creator/bo-creator.component';
import { InputGroupModule } from '@farris/ui-input-group';
import { FormsModule } from '@angular/forms';
import { InputsModule } from '@progress/kendo-angular-inputs';
import { NotifyModule } from '@farris/ui-notify';
import { ComboListModule } from '@farris/ui-combo-list';
import { BoDtServiceModule, BusinessObjectDtService } from '@gsp-lcm/bo-dt-service';
import { AppCreatorModalComponent } from './views/bo-creator/modal/app-modal/app-creator-modal.component';
import { BoCreatorUtils } from './views/bo-creator/utils/bo-create-utils';
import { SuCreatorModalComponent } from './views/bo-creator/modal/su-modal/su-creator-modal.component';
import { BoCreatorModalComponent } from './views/bo-creator/modal/bo-modal/bo-creator-modal.component';
var BOCreatorPackageModule = /** @class */ (function () {
    function BOCreatorPackageModule() {
    }
    BOCreatorPackageModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        BOCreatorComponent,
                        AppCreatorModalComponent,
                        SuCreatorModalComponent,
                        BoCreatorModalComponent
                    ],
                    imports: [
                        CommonModule,
                        InputGroupModule,
                        FormsModule,
                        InputsModule,
                        NotifyModule.forRoot(),
                        ComboListModule,
                        BoDtServiceModule
                    ],
                    entryComponents: [BOCreatorComponent],
                    exports: [BOCreatorComponent, AppCreatorModalComponent],
                    providers: [BusinessObjectDtService, BoCreatorUtils]
                },] }
    ];
    return BOCreatorPackageModule;
}());
export { BOCreatorPackageModule };
