// @farris/ui-filter-editor/esm5/lib/filter.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @Company: Inspur
 * @Version: v0.0.1
 * @Date: 2019-04-09 11:03:50
 * @LastEditTime: 2020-12-24 15:11:51
 */
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { LookupModule } from '@farris/ui-lookup';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { PerfectScrollbarModule } from '@farris/ui-perfect-scrollbar';
import { MessagerModule } from '@farris/ui-messager';
import { FarrisCommonModule } from '@farris/ui-common';
import { InputGroupModule } from '@farris/ui-input-group';
import { FilterComponent } from './filter.component';
import { FilterSqlCodeComponent } from './components/filter-sql.component';
import { FilterCodeComponent } from './components/filter-code.component';
import { FilterFooterComponent } from './components/filter-footer.component';
import { FilterTabsComponent } from './components/filter-tabs.component';
import { FilterEditorComponent } from './components/filter-editor.component';
import { FilterGridComponent } from './components/filter-grid.component';
import { LocaleModule } from '@farris/ui-locale';
import { ComboLookupModule } from '@farris/ui-combo-lookup';
import { FarrisFormsModule } from '@farris/ui-forms';
var FilterModule = /** @class */ (function () {
    function FilterModule() {
    }
    FilterModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        FarrisCommonModule,
                        MessagerModule.forRoot(),
                        PerfectScrollbarModule,
                        FarrisDialogModule,
                        LookupModule,
                        InputGroupModule,
                        LocaleModule.forRoot(),
                        ComboLookupModule,
                        FarrisFormsModule
                    ],
                    exports: [
                        FilterComponent,
                        FilterEditorComponent,
                        FilterGridComponent,
                        FilterFooterComponent
                    ],
                    declarations: [
                        FilterTabsComponent,
                        FilterGridComponent,
                        FilterFooterComponent,
                        FilterEditorComponent,
                        FilterCodeComponent,
                        FilterSqlCodeComponent,
                        FilterComponent
                    ],
                    providers: [],
                    entryComponents: [
                        FilterEditorComponent
                    ]
                },] }
    ];
    return FilterModule;
}());
export { FilterModule };
