// @gsp-wf/wf-process-editor/esm2015/lib/views/variable-setting/views/constant-edit/constant-edit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, TemplateRef, ViewChild, Input, Output, EventEmitter, ChangeDetectorRef, Optional, Inject, LOCALE_ID } from '@angular/core';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../../../i18n/pipe/wf-process-editor-locale.pipe';
export class ConstantEditComponent {
    /**
     * @param {?} cdr
     * @param {?} localeId
     * @param {?} resources
     * @param {?} localePipe
     */
    constructor(cdr, localeId, resources, localePipe) {
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.closeEmitter = new EventEmitter();
        this.submitEmitter = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    clickOK() {
        this.submitEmitter.emit(this.value);
    }
    /**
     * @return {?}
     */
    clickCancel() {
        this.closeEmitter.emit();
    }
}
ConstantEditComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-constant-edit',
                template: "<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <textarea class=\"form-control\" type=\"text\" value=\"message\" [(ngModel)]=\"value\"\r\n        name=\"message\" style=\"height: 100%;\"></textarea>\r\n      <div class=\"farris-feedback valid-feedback\"></div>\r\n    </section>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #footer>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm\" (click)=\"clickCancel()\">{{'component.optins.clickCancel' | wfLocale}}</button>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">{{'component.optins.clickOk' | wfLocale}}</button>\r\n</ng-template>\r\n",
                providers: [
                    WfLocalePipe,
                ],
                styles: [""]
            }] }
];
/** @nocollapse */
ConstantEditComponent.ctorParameters = () => [
    { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
    { type: WfLocalePipe, decorators: [{ type: Optional }] }
];
ConstantEditComponent.propDecorators = {
    footer: [{ type: ViewChild, args: ['footer',] }],
    closeEmitter: [{ type: Output }],
    submitEmitter: [{ type: Output }],
    value: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    ConstantEditComponent.prototype.footer;
    /** @type {?} */
    ConstantEditComponent.prototype.closeEmitter;
    /** @type {?} */
    ConstantEditComponent.prototype.submitEmitter;
    /** @type {?} */
    ConstantEditComponent.prototype.value;
    /**
     * @type {?}
     * @private
     */
    ConstantEditComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    ConstantEditComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    ConstantEditComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    ConstantEditComponent.prototype.localePipe;
}
