// @ecp-caf/common-structure/esm5/lib/convertor/constraint/constraint-convertor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ConstraintConst } from '../../define/constant/constraint-const';
var ConstraintConvertor = /** @class */ (function () {
    function ConstraintConvertor() {
    }
    /**
     * @param {?} obj
     * @param {?} constraint
     * @param {?} ctxt
     * @return {?}
     */
    ConstraintConvertor.prototype.convertToObject = /**
     * @param {?} obj
     * @param {?} constraint
     * @param {?} ctxt
     * @return {?}
     */
    function (obj, constraint, ctxt) {
        if (!constraint) {
            return;
        }
        switch (constraint.getKind()) {
            case ConstraintConst.DECIMAL:
                /** @type {?} */
                var decimalconstraint = (/** @type {?} */ (constraint));
                obj['scale'] = (/** @type {?} */ (decimalconstraint.scale));
                obj['pricision'] = (/** @type {?} */ (decimalconstraint.pricision));
            case ConstraintConst.STRING:
                /** @type {?} */
                var stringconstraint = (/** @type {?} */ (constraint));
                obj['maxLength'] = (/** @type {?} */ (stringconstraint.maxLength));
            default:
                throw new Error("\u7EA6\u675F\u4E0D\u652F\u6301\u7684" + constraint.getKind() + "\u7C7B\u578B\uFF01");
        }
    };
    /**
     * @param {?} constraint
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    ConstraintConvertor.prototype.initFromObject = /**
     * @param {?} constraint
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    function (constraint, obj, ctxt) {
        if (!obj) {
            return null;
        }
        switch (obj['kind']) {
            case ConstraintConst.DECIMAL:
            case ConstraintConst.STRING:
            default:
                throw new Error("\u7EA6\u675F\u4E0D\u652F\u6301\u7684" + obj['kind'] + "\u7C7B\u578B\uFF01");
        }
    };
    return ConstraintConvertor;
}());
export { ConstraintConvertor };
