// @gsp-wf/biz-process-api/fesm5/gsp-wf-biz-process-api.js
import { Injectable, NgModule, Component, defineInjectable } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BizProcessApiService = /** @class */ (function () {
    function BizProcessApiService() {
    }
    BizProcessApiService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    BizProcessApiService.ctorParameters = function () { return []; };
    /** @nocollapse */ BizProcessApiService.ngInjectableDef = defineInjectable({ factory: function BizProcessApiService_Factory() { return new BizProcessApiService(); }, token: BizProcessApiService, providedIn: "root" });
    return BizProcessApiService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BizProcessApiComponent = /** @class */ (function () {
    function BizProcessApiComponent() {
    }
    /**
     * @return {?}
     */
    BizProcessApiComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    BizProcessApiComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-biz-process-api',
                    template: "\n    <p>\n      biz-process-api works!\n    </p>\n  "
                }] }
    ];
    /** @nocollapse */
    BizProcessApiComponent.ctorParameters = function () { return []; };
    return BizProcessApiComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BizProcessApiModule = /** @class */ (function () {
    function BizProcessApiModule() {
    }
    BizProcessApiModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [BizProcessApiComponent],
                    imports: [],
                    exports: [BizProcessApiComponent]
                },] }
    ];
    return BizProcessApiModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { BizProcessApiService, BizProcessApiComponent, BizProcessApiModule };
