// @gsp-bef/gsp-vo-metadata/esm5/lib/view-model/src/VoDataExtendInfo/VoDataExtendInfo.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 数据逻辑扩展
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 数据逻辑扩展
 */
var /**
 * 数据逻辑扩展
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 数据逻辑扩展
 */
VoDataExtendInfo = /** @class */ (function () {
    function VoDataExtendInfo() {
    }
    return VoDataExtendInfo;
}());
/**
 * 数据逻辑扩展
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 数据逻辑扩展
 */
export { VoDataExtendInfo };
if (false) {
    /** @type {?} */
    VoDataExtendInfo.prototype.DataMappingActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.BeforeQueryActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.QueryActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.AfterQueryActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.BeforeRetrieveActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.RetrieveActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.AfterRetrieveActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.BeforeModifyActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.ModifyActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.AfterModifyActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.ChangesetMappingActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.BeforeCreateActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.CreateActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.AfterCreateActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.BeforeDeleteActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.DeleteActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.AfterDeleteActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.BeforeSaveActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.DataReversalMappingActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.AfterSaveActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.ChangesetReversalMappingActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.BeforeMultiDeleteActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.MultiDeleteActions;
    /** @type {?} */
    VoDataExtendInfo.prototype.AfterMultiDeleteActions;
}
/** @type {?} */
export var VoDataExtendInfos = [{
        Name: '扩展操作',
        Value: 'DataExtendInfo',
        items: [
            {
                Index: 0,
                Name: '数据Mapping',
                Value: 'DataMappingActions',
                HasInternalAction: true
            },
            {
                Index: 1,
                Name: '查询数据前',
                Value: 'BeforeQueryActions',
                HasInternalAction: false
            },
            {
                Index: 2,
                Name: '查询数据',
                Value: 'QueryActions',
                HasInternalAction: true
            },
            {
                Index: 3,
                Name: '查询数据后',
                Value: 'AfterQueryActions',
                HasInternalAction: false
            },
            {
                Index: 4,
                Name: '检索数据前',
                Value: 'BeforeRetrieveActions',
                HasInternalAction: false
            },
            {
                Index: 5,
                Name: '检索数据',
                Value: 'RetrieveActions',
                HasInternalAction: true
            },
            {
                Index: 6,
                Name: '检索数据后',
                Value: 'AfterRetrieveActions',
                HasInternalAction: false
            },
            {
                Index: 7,
                Name: '修改数据前',
                Value: 'BeforeModifyActions',
                HasInternalAction: false
            },
            {
                Index: 8,
                Name: '修改数据',
                Value: 'ModifyActions',
                HasInternalAction: true
            },
            {
                Index: 9,
                Name: '修改数据后',
                Value: 'AfterModifyActions',
                HasInternalAction: false
            },
            {
                Index: 10,
                Name: '变更集Mapping',
                Value: 'ChangesetMappingActions',
                HasInternalAction: true
            },
            {
                Index: 11,
                Name: '新增数据前',
                Value: 'BeforeCreateActions',
                HasInternalAction: false
            },
            {
                Index: 12,
                Name: '新增数据',
                Value: 'CreateActions',
                HasInternalAction: true
            },
            {
                Index: 13,
                Name: '新增数据后',
                Value: 'AfterCreateActions',
                HasInternalAction: false
            },
            {
                Index: 14,
                Name: '删除数据前',
                Value: 'BeforeDeleteActions',
                HasInternalAction: false
            },
            {
                Index: 15,
                Name: '删除数据',
                Value: 'DeleteActions',
                HasInternalAction: true
            },
            {
                Index: 16,
                Name: '删除数据后',
                Value: 'AfterDeleteActions',
                HasInternalAction: false
            },
            {
                Index: 17,
                Name: '批量删除数据前',
                Value: 'BeforeMultiDeleteActions',
                HasInternalAction: false
            },
            {
                Index: 18,
                Name: '批量删除数据',
                Value: 'MultiDeleteActions',
                HasInternalAction: true
            },
            {
                Index: 19,
                Name: '批量删除数据后',
                Value: 'AfterMultiDeleteActions',
                HasInternalAction: false
            },
            {
                Index: 20,
                Name: '保存数据前',
                Value: 'BeforeSaveActions',
                HasInternalAction: false
            },
            {
                Index: 21,
                Name: '数据反向Mapping',
                Value: 'DataReversalMappingActions',
                HasInternalAction: true
            },
            {
                Index: 22,
                Name: '保存数据后',
                Value: 'AfterSaveActions',
                HasInternalAction: false
            },
            {
                Index: 23,
                Name: '变更集反向Mapping',
                Value: 'ChangesetReversalMappingActions',
                HasInternalAction: true
            }
        ]
    }];
var VoDataExtendInfoData = /** @class */ (function () {
    function VoDataExtendInfoData() {
    }
    return VoDataExtendInfoData;
}());
if (false) {
    /** @type {?} */
    VoDataExtendInfoData.prototype.Index;
    /** @type {?} */
    VoDataExtendInfoData.prototype.Name;
    /** @type {?} */
    VoDataExtendInfoData.prototype.Value;
    /** @type {?} */
    VoDataExtendInfoData.prototype.HasInternalAction;
    /** @type {?} */
    VoDataExtendInfoData.prototype.items;
}
