// @farris/ui-language-textbox/esm5/lib/language-textbox.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LocaleModule } from '@farris/ui-locale';
import { InputGroupModule } from '@farris/ui-input-group';
import { LanguageTextboxComponent } from './language-textbox.component';
import { LanguageTextPanelComponent } from './language-textbox-panel.component';
var LanguageTextboxModule = /** @class */ (function () {
    function LanguageTextboxModule() {
    }
    LanguageTextboxModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        LanguageTextboxComponent,
                        LanguageTextPanelComponent
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        InputGroupModule,
                        LocaleModule.forRoot()
                    ],
                    exports: [LanguageTextboxComponent],
                    providers: [],
                    entryComponents: [LanguageTextPanelComponent]
                },] }
    ];
    return LanguageTextboxModule;
}());
export { LanguageTextboxModule };
