// @farris/dynamic-control-group/esm2015/lib/components/date/single-date.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from "@angular/core";
import { FormGroup } from "@angular/forms";
export class SingleDateComponent {
    constructor() { }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.control.control.format = this.control.control.format || 'yyyy-MM-dd';
        this.className = this.control.className ? this.control.className : 'col-12 col-md-6 col-xl-3 col-el-2';
    }
}
SingleDateComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-single-date",
                template: `
    <div class="farris-group-wrap" [formGroup]="form">
      <div class="form-group farris-form-group">
        <label class="col-form-label">
          <span *ngIf="0" class="farris-label-info text-danger">*</span>
          <span class="farris-label-text">{{control.name}}</span>
        </label>
        <div class="farris-input-wrap">
          <farris-datepicker [returnFormat]="control.control.format" [formControlName]="control.labelCode" [readonly]="control.control.readonly"></farris-datepicker>
        </div>
      </div>
    </div>
  `
            }] }
];
/** @nocollapse */
SingleDateComponent.ctorParameters = () => [];
SingleDateComponent.propDecorators = {
    className: [{ type: HostBinding, args: ['class',] }],
    control: [{ type: Input }],
    form: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    SingleDateComponent.prototype.className;
    /** @type {?} */
    SingleDateComponent.prototype.control;
    /** @type {?} */
    SingleDateComponent.prototype.form;
}
