// @farris/mobile-command-services/esm2015/lib/ui-services/dialog.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ui-services/dialog.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { from } from 'rxjs';
import { Dialog } from '@farris/mobile-ui';
class DialogService {
    /**
     * 提示对话框
     * \@summary
     * type=info|success|warning|error
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    alert(message, options) {
        /** @type {?} */
        const promise$ = new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        (resolve, reject) => {
            Dialog.alert(Object.assign({ title: '提醒', message: message, cancelText: '取消', confirmText: '确定', onConfirm: (/**
                 * @return {?}
                 */
                () => { resolve(true); }), onCancel: (/**
                 * @return {?}
                 */
                () => { resolve(false); }) }, options));
        }));
        return from(promise$);
    }
    /**
     * 确认框对话框
     * @param {?} message
     * @return {?}
     */
    confirm(message) {
        /** @type {?} */
        const confirmResult$ = this.confirmPromise(message);
        return from(confirmResult$);
    }
    /**
     * 自定义弹窗
     * @param {?} options
     * @return {?}
     */
    show(options) {
        this.confirmPromise(options);
    }
    /**
     * 输入对话框
     * @param {?} title 标题
     * @param {?=} options 配置信息
     * @return {?}
     */
    prompt(title, options) {
        /** @type {?} */
        const promise$ = new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        (resolve, reject) => {
            /** @type {?} */
            const vm = Dialog.prompt(Object.assign({ title: title, confirmText: '确定', onConfirm: (/**
                 * @return {?}
                 */
                () => {
                    resolve(vm.$data.promptText);
                }), onCancel: (/**
                 * @return {?}
                 */
                () => { resolve('取消'); }), onShow: (/**
                 * @return {?}
                 */
                () => {
                    console.log('显示弹窗');
                }) }, options));
        }));
        return from(promise$);
    }
    /**
     * 使用Promise实例包装回调方法，以同步回调方法执行后结果
     * @private
     * @param {?=} message
     * @return {?}
     */
    confirmPromise(message) {
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        (resolve, reject) => {
            Dialog.confirm({
                message: message,
                onConfirm: (/**
                 * @return {?}
                 */
                () => { resolve(true); }),
                onCancel: (/**
                 * @return {?}
                 */
                () => { resolve(false); }),
            });
        }));
    }
}
export { DialogService };
