// @farris/mobile-command-services/esm2015/lib/ui-services/notify.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ui-services/notify.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Notify } from '@farris/mobile-ui';
/**
 * 消息提示服务
 * \@summary
 * 页面边缘滑出，自动消失，语气略强于Toast
 */
export class NotifyService {
    /**
     * 消息提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    info(message, options) {
        ((/** @type {?} */ (Notify))).info(message, Object.assign({ duration: 1500 }, options));
    }
    /**
     * 成功提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    success(message, options) {
        ((/** @type {?} */ (Notify))).success(message, Object.assign({ duration: 1000 }, options));
    }
    /**
     * 警告提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    warning(message, options) {
        ((/** @type {?} */ (Notify))).warning(message, Object.assign({ duration: 1500 }, options));
    }
    /**
     * 失败提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    error(message, options) {
        ((/** @type {?} */ (Notify))).error(message, Object.assign({ duration: 1500 }, options));
    }
    /**
     * 自定义提醒
     * @param {?} options
     * @return {?}
     */
    show(options) {
        ((/** @type {?} */ (Notify))).show(options);
    }
}
