// @farris/mobile-command-services/esm2015/lib/ui-services/toast.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ui-services/toast.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Toast } from '@farris/mobile-ui';
/**
 * 轻提示服务
 */
class ToastService {
    /**
     * 消息提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    info(message, options) {
        ((/** @type {?} */ (Toast))).info(message, options);
    }
    /**
     * 成功提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    success(message, options) {
        ((/** @type {?} */ (Toast))).success(message, options);
    }
    /**
     * 警告提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    warning(message, options) {
        ((/** @type {?} */ (Toast))).warning(message, options);
    }
    /**
     * 失败提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    error(message, options) {
        ((/** @type {?} */ (Toast))).error(message, options);
    }
    /**
     * 自定义提示
     * @param {?=} options
     * @return {?}
     */
    show(options) {
        ((/** @type {?} */ (Toast)))(options);
    }
}
export { ToastService };
