// @farris/ui-datagrid-settings/esm5/lib/datagrid-sort-setting.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, Injector, EventEmitter, Output, ChangeDetectorRef } from '@angular/core';
import { MultiSelectComponent } from '@farris/ui-multi-select';
var DatagridSortSettingComponent = /** @class */ (function () {
    function DatagridSortSettingComponent(injector, cd) {
        this.injector = injector;
        this.cd = cd;
        this.idField = 'field';
        this.valueField = 'field';
        this.textField = 'title';
        this.selectedIds = [];
        this.treeData = [];
        this.sortInfoChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    DatagridSortSettingComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} $event
     * @param {?} data
     * @param {?} order
     * @return {?}
     */
    DatagridSortSettingComponent.prototype.setFieldSortOrder = /**
     * @param {?} $event
     * @param {?} data
     * @param {?} order
     * @return {?}
     */
    function ($event, data, order) {
        $event.stopPropagation();
        data.sortable = true;
        data.order = order;
        if (!order) {
            this.ms.moveToLeft(data);
        }
        this.cd.detectChanges();
        this.onSelectionsChange(null);
        return false;
    };
    /**
     * @return {?}
     */
    DatagridSortSettingComponent.prototype.getSortInfo = /**
     * @return {?}
     */
    function () {
        if (this.ms.selections) {
            /** @type {?} */
            var names_1 = [];
            /** @type {?} */
            var orders_1 = [];
            this.ms.selections.forEach((/**
             * @param {?} col
             * @return {?}
             */
            function (col) {
                col.order = col.order || 'asc';
                names_1.push(col.field);
                orders_1.push(col.order);
            }));
            return { sortName: names_1, sortOrder: orders_1 };
        }
        return '';
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridSortSettingComponent.prototype.onSelectionsChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var si = this.getSortInfo();
        this.sortInfoChange.emit(si);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridSortSettingComponent.prototype.onBeforeMove = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var items = (/** @type {?} */ ($event.items));
        if ($event.target.indexOf('rightAll') > -1) {
            for (var id in items) {
                if ($event.items[id]) {
                    $event.items[id].order = 'asc';
                }
            }
        }
        else {
            if ($event.target === 'right') {
                for (var id in items) {
                    if ($event.items[id]) {
                        $event.items[id].order = 'asc';
                    }
                }
            }
        }
    };
    DatagridSortSettingComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sort-setting',
                    template: "<div style=\"padding: 10px;height: 100%;\">\r\n<multi-select [dataSource]=\"treeData\" [idField]=\"idField\" [valueField]=\"valueField\"\r\n[textField]=\"textField\" [displayType]=\"'List'\" style=\"height: 100%\" [height]=\"0\" [showCheckbox]=\"false\"\r\n[(selectedId)]=\"selectedIds\" [leftTitle]=\"'datagrid.settings.canchoose'|locale\" [rightTitle]=\"'datagrid.settings.choosed'|locale\" #ms (dataChange)=\"onSelectionsChange($event)\" (beforeMove)=\"onBeforeMove($event)\">\r\n    <ng-template farrisTemplate=\"text\" let-data let-side=\"side\" let-selected=\"selected\" >\r\n        <ng-container *ngIf=\"side === 'L'\">\r\n            {{ data[textField] }}\r\n        </ng-container>\r\n        <ng-container *ngIf=\"side === 'R' \">\r\n            <!-- <span class=\"f-icon f-icon-sort-asc-sm sort-label\" title=\"\u5347\u5E8F\" *ngIf=\"data.sortable && data.order === 'asc'\"></span>\r\n            <span class=\"f-icon f-icon-sort-desc-sm  sort-label\" title=\"\u964D\u5E8F\" *ngIf=\"data.sortable && data.order === 'desc'\"></span> -->\r\n            <div class=\"d-flex\">\r\n                <div style=\"width: 120px;white-space: nowrap;min-width: 60px\">\r\n                    {{ data[textField] }} \r\n                </div>\r\n                <div class=\"farris-input-wrap  farris-checkradio-hor flex-fill\" style=\"justify-content: right;flex-wrap:nowrap\">\r\n                    <button type=\"button\" class=\"btn btn-link btn-icontext mr-2\" style=\"font-size: 12px;\" (click)=\"setFieldSortOrder($event, data, 'asc')\">\r\n                        <span class=\"f-icon\" [class.f-icon-radiobutton]=\" data.order && data.order !== 'asc'\" [class.f-icon-check-circle]=\"!data.order || data.order === 'asc'\"></span>\r\n                        {{'datagrid.settings.asc'|locale}}\r\n                    </button>\r\n                    \r\n                    <button type=\"button\" class=\"btn btn-link btn-icontext mr-2\" style=\"font-size: 12px;\" (click)=\"setFieldSortOrder($event, data, 'desc')\">\r\n                        <span class=\"f-icon\" [class.f-icon-radiobutton]=\"data.order !== 'desc'\" [class.f-icon-check-circle]=\"data.order === 'desc'\"></span>\r\n                        {{'datagrid.settings.desc'|locale}}\r\n                    </button>\r\n\r\n                    <button type=\"button\" class=\"btn btn-link btn-icontext sort-btns\" style=\"font-size: 12px;\" style=\"margin-right: 0;\" (click)=\"setFieldSortOrder($event, data, '')\" title=\"{{'datagrid.settings.cancelSort'|locale}}\">\r\n                        <span class=\"f-icon  f-icon-unsort\"></span>\r\n                    </button>\r\n                </div>\r\n            </div>\r\n\r\n        </ng-container>\r\n    </ng-template>\r\n</multi-select>\r\n</div>"
                }] }
    ];
    /** @nocollapse */
    DatagridSortSettingComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ChangeDetectorRef }
    ]; };
    DatagridSortSettingComponent.propDecorators = {
        idField: [{ type: Input }],
        valueField: [{ type: Input }],
        textField: [{ type: Input }],
        selectedIds: [{ type: Input }],
        treeData: [{ type: Input }],
        ms: [{ type: ViewChild, args: ['ms',] }],
        sortInfoChange: [{ type: Output }]
    };
    return DatagridSortSettingComponent;
}());
export { DatagridSortSettingComponent };
if (false) {
    /** @type {?} */
    DatagridSortSettingComponent.prototype.idField;
    /** @type {?} */
    DatagridSortSettingComponent.prototype.valueField;
    /** @type {?} */
    DatagridSortSettingComponent.prototype.textField;
    /** @type {?} */
    DatagridSortSettingComponent.prototype.selectedIds;
    /** @type {?} */
    DatagridSortSettingComponent.prototype.treeData;
    /** @type {?} */
    DatagridSortSettingComponent.prototype.ms;
    /** @type {?} */
    DatagridSortSettingComponent.prototype.sortInfoChange;
    /**
     * @type {?}
     * @private
     */
    DatagridSortSettingComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    DatagridSortSettingComponent.prototype.cd;
}
