// @farris/ui-datagrid-settings/esm5/lib/datagrid-scheme.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { appendPerfixForUri, IdService } from "@farris/ui-common";
import { of, Subject } from "rxjs";
import { IS_MULTI_HEADER_STRING } from "./setting.model";
import { cloneDeep } from "lodash-es";
/**
 * 保存方案API
 * @type {?}
 */
var SCHEME_WEBAPI = '/api/runtime/sys/v1.0/querysolution';
/**
 * 获取方案列表API
 * @type {?}
 */
var SCHEME_WEBAPI_QUERY = SCHEME_WEBAPI + "/belongId/";
/**
 * 方案列表管理- 设默认、删除 API
 * @type {?}
 */
var SCHEME_WEBAPI_UPDATE = SCHEME_WEBAPI + "/batch";
/**
 * 权限验证
 * @type {?}
 */
var SCHEME_WEBAPI_Auth = '/api/runtime/sys/v1.0/querysolution/componentType/Datagrid';
/** @type {?} */
var LANGUAGE_WEBAPI = '/api/runtime/sys/v1.0/loginInfo?infoType=supportedLanguage';
var DatagridSchemeService = /** @class */ (function () {
    function DatagridSchemeService(inject) {
        this.inject = inject;
        this.restService = null;
        this.state = {};
        this.perfixStr = '';
        this.schemeList$ = new Subject();
        this.schemeApi = {
            SCHEME_WEBAPI: SCHEME_WEBAPI,
            /** 获取方案列表API */
            SCHEME_WEBAPI_QUERY: SCHEME_WEBAPI_QUERY,
            /** 方案列表管理- 设默认、删除 API */
            SCHEME_WEBAPI_UPDATE: SCHEME_WEBAPI_UPDATE,
            /** 权限验证 */
            SCHEME_WEBAPI_Auth: SCHEME_WEBAPI_Auth,
            LANGUAGE_WEBAPI: LANGUAGE_WEBAPI
        };
        this.schemeKey = {};
        this.currentSchemeSettings = null;
    }
    /**
     * @param {?} perfixStr
     * @return {?}
     */
    DatagridSchemeService.prototype.setUriPerfix = /**
     * @param {?} perfixStr
     * @return {?}
     */
    function (perfixStr) {
        this.perfixStr = perfixStr || '';
    };
    /**
     * @param {?} d
     * @param {?} gridId
     * @return {?}
     */
    DatagridSchemeService.prototype.update = /**
     * @param {?} d
     * @param {?} gridId
     * @return {?}
     */
    function (d, gridId) {
        if (!this.state[gridId]) {
            this.state[gridId] = {};
        }
        this.state[gridId] = Object.assign(this.state[gridId], d);
    };
    /**
     * @param {?} httpSer
     * @return {?}
     */
    DatagridSchemeService.prototype.setRestService = /**
     * @param {?} httpSer
     * @return {?}
     */
    function (httpSer) {
        if (httpSer && httpSer['befRepository']) {
            this.restService = httpSer['befRepository']['restService'];
        }
    };
    /**
     * @private
     * @param {?} uri
     * @param {?} perfix
     * @return {?}
     */
    DatagridSchemeService.prototype.mergeUri = /**
     * @private
     * @param {?} uri
     * @param {?} perfix
     * @return {?}
     */
    function (uri, perfix) {
        return appendPerfixForUri ? appendPerfixForUri(uri, perfix) : uri;
    };
    /**
     * @private
     * @return {?}
     */
    DatagridSchemeService.prototype.getWebFormKey = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var webformHash = window.location.hash.split('?')[0];
        return webformHash.substring(webformHash.lastIndexOf('/') + 1);
    };
    /**
     * @param {?} gridId
     * @return {?}
     */
    DatagridSchemeService.prototype.destory = /**
     * @param {?} gridId
     * @return {?}
     */
    function (gridId) {
        delete this.schemeKey[gridId];
    };
    /**
     * @param {?} gridId
     * @return {?}
     */
    DatagridSchemeService.prototype.getSchemeKey = /**
     * @param {?} gridId
     * @return {?}
     */
    function (gridId) {
        if (!this.schemeKey[gridId]) {
            /** @type {?} */
            var formKey = this.getWebFormKey();
            this.schemeKey[gridId] = formKey + "_DatagridScheme_" + gridId;
        }
        return this.schemeKey[gridId];
    };
    /**
     * @param {?} gridID
     * @param {?=} isMultiHeader
     * @return {?}
     */
    DatagridSchemeService.prototype.getSchemeList = /**
     * @param {?} gridID
     * @param {?=} isMultiHeader
     * @return {?}
     */
    function (gridID, isMultiHeader) {
        if (isMultiHeader === void 0) { isMultiHeader = false; }
        /** @type {?} */
        var uri = this.schemeApi.SCHEME_WEBAPI_QUERY + this.getBelongId(gridID, isMultiHeader);
        uri = this.mergeUri(uri, this.perfixStr);
        if (this.restService) {
            return this.restService.invoke(uri, 'GET', null, null, false);
        }
    };
    /**
     * @param {?} gridID
     * @param {?} schemeData
     * @return {?}
     */
    DatagridSchemeService.prototype.updateSchemeData = /**
     * @param {?} gridID
     * @param {?} schemeData
     * @return {?}
     */
    function (gridID, schemeData) {
        /** @type {?} */
        var schemeList = this.getStateValue(gridID, 'list');
        schemeList = schemeList.filter((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n.id !== schemeData.id; }));
        schemeList.push(schemeData);
        this.setSchemeList(gridID, schemeList);
    };
    /**
     * @private
     * @param {?} gridID
     * @param {?=} isMultiHeader
     * @return {?}
     */
    DatagridSchemeService.prototype.getBelongId = /**
     * @private
     * @param {?} gridID
     * @param {?=} isMultiHeader
     * @return {?}
     */
    function (gridID, isMultiHeader) {
        if (isMultiHeader === void 0) { isMultiHeader = false; }
        /** @type {?} */
        var id = this.getSchemeKey(gridID);
        if (isMultiHeader) {
            id = id + IS_MULTI_HEADER_STRING;
        }
        return id;
    };
    /**
     * @param {?} scheme
     * @param {?} gridID
     * @param {?=} isUpdate
     * @param {?=} isMultiHeader
     * @return {?}
     */
    DatagridSchemeService.prototype.saveScheme = /**
     * @param {?} scheme
     * @param {?} gridID
     * @param {?=} isUpdate
     * @param {?=} isMultiHeader
     * @return {?}
     */
    function (scheme, gridID, isUpdate, isMultiHeader) {
        if (isUpdate === void 0) { isUpdate = false; }
        if (isMultiHeader === void 0) { isMultiHeader = false; }
        if (this.restService) {
            /** @type {?} */
            var httpMethod = isUpdate ? 'PUT' : 'POST';
            scheme.belongId = this.getBelongId(gridID, isMultiHeader);
            /** @type {?} */
            var uri = this.mergeUri(this.schemeApi.SCHEME_WEBAPI, this.perfixStr);
            return this.restService.invoke(uri, httpMethod, null, { body: scheme }, false);
        }
    };
    /**
     * @param {?} param
     * @param {?} gridID
     * @param {?=} isMultiHeader
     * @return {?}
     */
    DatagridSchemeService.prototype.updateScheme = /**
     * @param {?} param
     * @param {?} gridID
     * @param {?=} isMultiHeader
     * @return {?}
     */
    function (param, gridID, isMultiHeader) {
        if (isMultiHeader === void 0) { isMultiHeader = false; }
        if (!param) {
            return of(false);
        }
        /** @type {?} */
        var belongId = this.getBelongId(gridID, isMultiHeader);
        param.belongId = belongId;
        if (param.belongId) {
            /** @type {?} */
            var uri = this.mergeUri(this.schemeApi.SCHEME_WEBAPI_UPDATE, this.perfixStr);
            return this.restService.invoke(uri, 'PUT', null, { body: param }, false);
        }
    };
    /**
     * @param {?} gridId
     * @param {...?} statePro
     * @return {?}
     */
    DatagridSchemeService.prototype.getStateValue = /**
     * @param {?} gridId
     * @param {...?} statePro
     * @return {?}
     */
    function (gridId) {
        var statePro = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            statePro[_i - 1] = arguments[_i];
        }
        /** @type {?} */
        var dgState = this.state[gridId];
        if (dgState) {
            if (statePro && statePro.length) {
                return statePro.reduce((/**
                 * @param {?} r
                 * @param {?} c
                 * @return {?}
                 */
                function (r, c) {
                    return r[c];
                }), dgState);
            }
            return dgState;
        }
        return null;
    };
    /**
     * @private
     * @param {?} gridId
     * @param {?} propertyName
     * @param {?} value
     * @return {?}
     */
    DatagridSchemeService.prototype.updateStateValue = /**
     * @private
     * @param {?} gridId
     * @param {?} propertyName
     * @param {?} value
     * @return {?}
     */
    function (gridId, propertyName, value) {
        var _a;
        this.update((_a = {}, _a[propertyName] = value, _a), gridId);
    };
    /**
     * @param {?} gridId
     * @param {?} newSchemeList
     * @return {?}
     */
    DatagridSchemeService.prototype.setSchemeList = /**
     * @param {?} gridId
     * @param {?} newSchemeList
     * @return {?}
     */
    function (gridId, newSchemeList) {
        this.updateStateValue(gridId, 'list', newSchemeList);
        this.schemeList$.next(newSchemeList);
    };
    /**
     * @param {?} gridId
     * @param {?} schemeName
     * @return {?}
     */
    DatagridSchemeService.prototype.hasSchemeName = /**
     * @param {?} gridId
     * @param {?} schemeName
     * @return {?}
     */
    function (gridId, schemeName) {
        /** @type {?} */
        var schemeList = this.getStateValue(gridId, 'list');
        if (!schemeList || !schemeList.length) {
            return false;
        }
        if (typeof schemeName === 'string') {
            return !!schemeList.find((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return n.name === schemeName.trim(); }));
        }
        else {
            if (typeof schemeName === 'object') {
                /** @type {?} */
                var replayNames_1 = [];
                schemeList.forEach((/**
                 * @param {?} n
                 * @return {?}
                 */
                function (n) {
                    if (n.name) {
                        /** @type {?} */
                        var nameObj = JSON.parse(n.name);
                        // const currentNames = Object.values(nameObj);
                        for (var k in schemeName) {
                            if (nameObj[k] === schemeName[k]) {
                                replayNames_1.push(k);
                            }
                        }
                    }
                }));
                return replayNames_1;
            }
        }
    };
    /**
     * @return {?}
     */
    DatagridSchemeService.prototype.checkAuthority = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var uri = this.mergeUri(this.schemeApi.SCHEME_WEBAPI_Auth, this.perfixStr);
        return this.restService.invoke(uri, 'GET', null, null, false);
    };
    /**
     * @return {?}
     */
    DatagridSchemeService.prototype.getLanguages = /**
     * @return {?}
     */
    function () {
        if (this.restService) {
            /** @type {?} */
            var uri = this.mergeUri(this.schemeApi.LANGUAGE_WEBAPI, this.perfixStr);
            return this.restService.invoke(uri, 'GET', null, null, false);
        }
        else {
            return of([]);
        }
    };
    /**
     * @return {?}
     */
    DatagridSchemeService.prototype.createSchemeTaskID = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var idService = new IdService();
        this['SCHEME_TASK_ID'] = idService.guid();
        this.currentSchemeSettings = null;
    };
    /**
     * @param {?} configs
     * @return {?}
     */
    DatagridSchemeService.prototype.setSettings = /**
     * @param {?} configs
     * @return {?}
     */
    function (configs) {
        this.currentSchemeSettings = {
            taskid: this['SCHEME_TASK_ID'],
            values: cloneDeep(configs)
        };
    };
    /**
     * @return {?}
     */
    DatagridSchemeService.prototype.getSettings = /**
     * @return {?}
     */
    function () {
        return this.currentSchemeSettings;
    };
    /**
     * @param {?} nowSettings
     * @param {?} isMultiHeader
     * @return {?}
     */
    DatagridSchemeService.prototype.isSchemeChanged = /**
     * @param {?} nowSettings
     * @param {?} isMultiHeader
     * @return {?}
     */
    function (nowSettings, isMultiHeader) {
        var sortInfo = nowSettings.sortInfo, viewColumns = nowSettings.viewColumns, groupField = nowSettings.groupField, columnFormat = nowSettings.columnFormat, expandGroupRows = nowSettings.expandGroupRows, multiHeaders = nowSettings.multiHeaders;
        var _a = this.getSettings().values, sortInfo1 = _a.sortInfo, viewColumns1 = _a.viewColumns, groupField1 = _a.groupField, columnFormat1 = _a.columnFormat, expandGroupRows1 = _a.expandGroupRows, multiHeaders1 = _a.multiHeaders;
        /** @type {?} */
        var checkSortInfo = (/**
         * @param {?} _sortInfo
         * @return {?}
         */
        function (_sortInfo) {
            if (!_sortInfo || !Object.keys(_sortInfo).length) {
                _sortInfo = {
                    sortName: [],
                    sortOrder: []
                };
            }
            return _sortInfo;
        });
        sortInfo = checkSortInfo(sortInfo);
        sortInfo1 = checkSortInfo(sortInfo1);
        /** @type {?} */
        var checkGroupField = (/**
         * @param {?} _groupField
         * @return {?}
         */
        function (_groupField) {
            if (!_groupField) {
                _groupField = [];
            }
            return _groupField;
        });
        groupField = checkGroupField(groupField);
        groupField1 = checkGroupField(groupField1);
        /** @type {?} */
        var nowconfig;
        /** @type {?} */
        var initConfig;
        if (isMultiHeader) {
            nowconfig = { sortInfo: sortInfo, expandGroupRows: expandGroupRows, groupField: groupField, multiHeaders: multiHeaders };
            initConfig = { sortInfo: sortInfo1, expandGroupRows: expandGroupRows1, groupField: groupField1, multiHeaders: multiHeaders1 };
        }
        else {
            nowconfig = { columnFormat: columnFormat, expandGroupRows: expandGroupRows, groupField: groupField, sortInfo: sortInfo, viewColumns: viewColumns };
            initConfig = {
                columnFormat: columnFormat1,
                expandGroupRows: expandGroupRows1,
                groupField: groupField1,
                sortInfo: sortInfo1,
                viewColumns: viewColumns1
            };
        }
        /** @type {?} */
        var _clearData = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            return JSON.stringify(data).replace(/UNFIXED/g, '').replace(/"calculationType":-1,/g, '').replace(/"calculationType":-1/g, '');
        });
        return _clearData(nowconfig) != _clearData(initConfig);
    };
    return DatagridSchemeService;
}());
export { DatagridSchemeService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DatagridSchemeService.prototype.restService;
    /**
     * @type {?}
     * @private
     */
    DatagridSchemeService.prototype.state;
    /**
     * @type {?}
     * @private
     */
    DatagridSchemeService.prototype.perfixStr;
    /** @type {?} */
    DatagridSchemeService.prototype.schemeList$;
    /**
     * @type {?}
     * @private
     */
    DatagridSchemeService.prototype.schemeApi;
    /**
     * @type {?}
     * @private
     */
    DatagridSchemeService.prototype.schemeKey;
    /**
     * @type {?}
     * @private
     */
    DatagridSchemeService.prototype.currentSchemeSettings;
    /**
     * @type {?}
     * @private
     */
    DatagridSchemeService.prototype.inject;
}
