// @farris/ui-datagrid-settings/esm5/lib/datagrid-settings-locale.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 列设置相关多语资源
 * @type {?}
 */
export var DATAGRID_SETTINGS_LOCALEDATA = {
    en: {
        'datagrid-settings': {
            noAuth: {
                public: 'You do not have permission to set the usage scope of the scheme as public.',
                org: 'You do not have permission to set the usage range of the scheme as organization.'
            },
            has: {
                exist: 'The scheme name already exists, please modify it.',
                chs: 'Simplified Chinese name already exists.',
                cht: 'Traditional Chinese name already exists.',
                en: 'English name already exists.'
            },
            required: {
                schemeName: 'Please enter a scheme name.',
                chs: 'Please enter a simplified Chinese name.',
                cht: 'Please enter a name in Traditional Chinese.',
                en: 'Please enter an English name.',
                org: 'Please select an organization.'
            },
            groupRowStatus: {
                label: 'Group Row Status',
                expand: 'open',
                collapse: 'fold'
            },
            column: {
                title: 'Column title',
                fixed: 'Fixed type',
                unFixed: 'Unfixed',
                leftFixed: 'Left fixed',
                rightFixed: 'Right fixed'
            }
        },
        'datagrid-scheme': {
            saveTitle: 'Save Scheme',
            namePlaceholder: 'Please enter a scheme name (no more than 9 characters)',
            setPrivate: 'Private',
            privateTip: 'When set as private, you can only view and use the scheme yourself',
            setPublic: 'Public',
            publicTip: 'When public, all users can view and use the scheme',
            setOrg: 'Organization',
            orgTip: 'When it is set to organization, users in the organization can view and use the scheme',
            setDefault: 'Default',
            defaultTip: 'The next time you enter this function, the scheme will be loaded by default',
            syspreset: 'System preset',
            isDefault: 'Default',
            isPublic: 'Public',
            isOrg: 'Org',
            isPrivate: 'Private',
            newScheme: 'Add New Scheme',
            mgrScheme: 'Manage',
            saveas: 'Save as',
            onlyshowcolumns: 'only the display column',
            schemeList: 'Scheme list',
            delete: 'Delete this shceme',
            columns: {
                name: 'Name',
                type: 'Type',
                isDefault: 'IsDefault',
                username: 'Creator',
                mgr: 'Manage'
            },
            currentSheme: 'Current scheme：',
            schemeChanged: {
                title: 'The scheme has been changed. Do you want to save it as a new scheme?',
                tips: 'cancel: discard modifications; No: save the current scheme; Yes: save as a new scheme.',
                yes: 'Yes',
                no: 'No',
                cancel: 'Cancel'
            },
            noDisplyColumnTip: 'Please set the fields to be displayed.'
        },
        'headerGroup': {
            title: 'Multiple meter heads',
            preview: 'Preview effect',
            tips: 'The fields in the (candidate column) will not be displayed in the list; Click on the field name and display the relevant operation buttons on the right.',
            contextMenus: {
                new: 'New Header',
                newChild: 'New Child Header',
                setDataField: 'Assign Data Columns',
                rename: 'Rename',
                top: 'Top',
                up: 'Up',
                down: 'Down',
                bottom: 'Bottom',
                cancelGroup: 'Cancel',
                remove: 'Remove'
            },
            candidate: 'Candidate columns',
            setHeader: 'Set as header',
            ok: 'Ok',
            cancel: 'Cancel',
            required: 'Please enter a header name!',
            defaultGroupHeaderName: 'NewHeader',
            newHeaderPlaceholder: 'Please enter a header name.',
            newHeaderLabel: 'Header Name'
        }
    },
    'zh-CHS': {
        'datagrid-settings': {
            noAuth: {
                public: '您没有将方案的使用范围设置为公共的权限。',
                org: '您没有将方案的使用范围设置为组织的权限。'
            },
            has: {
                exist: '方案名称已存在，请修改。',
                chs: '简体中文名称已存在。',
                cht: '繁体中文名称已存在。',
                en: '英文名称已存在。'
            },
            required: {
                schemeName: '请输入方案名称。',
                chs: '请输入简体中文名称。',
                cht: '请输入繁体中文名称。',
                en: '请输入英文名称。',
                org: '请选择组织。'
            },
            groupRowStatus: {
                label: '分组行状态',
                expand: '展开',
                collapse: '折叠'
            },
            column: {
                title: '列标题',
                fixed: '固定列',
                unFixed: '不固定',
                leftFixed: '固定在左侧',
                rightFixed: '固定在右侧'
            }
        },
        'datagrid-scheme': {
            saveTitle: '保存方案',
            namePlaceholder: '请输入方案名称（最多不超过9个字符）',
            setPrivate: '设为个人',
            privateTip: '设为个人时，只能自己查看、使用该方案',
            setPublic: '设为公共',
            publicTip: '设为公共时，所有用户都可查看、使用该方案',
            setOrg: '设为组织',
            orgTip: '设为组织时，组织内的用户均能查看、使用该方案',
            setDefault: '设为默认',
            defaultTip: '下次进入此功能时，默认加载该方案',
            syspreset: '系统预制',
            isDefault: '默认',
            isPublic: '公共',
            isOrg: '组织',
            isPrivate: '个人',
            newScheme: '新增方案',
            mgrScheme: '管理',
            saveas: '另存为',
            onlyshowcolumns: '仅看展示列',
            schemeList: '方案列表',
            delete: '删除',
            columns: {
                name: '名称',
                type: '属性',
                isDefault: '默认',
                username: '创建人',
                mgr: '管理'
            },
            currentSheme: '当前方案：',
            schemeChanged: {
                title: '方案已更改，是否另存为新方案？',
                tips: '取消：放弃修改；否：保存当前方案；是：方案另存为',
                yes: '是',
                no: '否',
                cancel: '取消'
            },
            noDisplyColumnTip: '请设置要显示的字段'
        },
        'headerGroup': {
            title: '多表头',
            preview: '预览效果',
            tips: '在（候选列）中的字段，将不在列表中显示; 点击字段名称，在右侧显示相关操作按钮。',
            contextMenus: {
                new: '新建表头',
                newChild: '新建子级表头',
                setDataField: '分配数据列',
                rename: '重命名',
                top: '置顶',
                up: '上移',
                down: '下移',
                bottom: '置底',
                cancelGroup: '取消分组',
                remove: '移除'
            },
            candidate: '(候选列)',
            setHeader: '设为表头',
            ok: '确定',
            cancel: '取消',
            required: '请输入表头名称！',
            defaultGroupHeaderName: '分组表头1',
            newHeaderPlaceholder: '请输入名称',
            newHeaderLabel: '表头名称'
        }
    },
    'zh-CHT': {
        'datagrid-settings': {
            noAuth: {
                public: '您沒有將方案的使用範圍設置爲公共的權限。',
                org: '您沒有將方案的使用範圍設置爲組織的權限。'
            },
            has: {
                exist: '方案名稱已存在，請修改。',
                chs: '簡體中文名稱已存在。',
                cht: '繁體中文名稱已存在。',
                en: '英文名稱已存在。'
            },
            required: {
                schemeName: '請輸入方案名稱.',
                chs: '請輸入簡體中文名稱。',
                cht: '請輸入繁體中文名稱。',
                en: '請輸入英文名稱。',
                org: '請選擇組織。'
            },
            groupRowStatus: {
                label: '分組行狀態',
                expand: '展開',
                collapse: '折疊'
            },
            column: {
                title: '列標題',
                fixed: '固定列',
                unFixed: '不固定',
                leftFixed: '固定在左側',
                rightFixed: '固定在右側'
            }
        },
        'datagrid-scheme': {
            saveTitle: '保存方案',
            namePlaceholder: '請輸入方案名稱（最多不超過9個字符）',
            setPrivate: '設爲個人',
            privateTip: '設爲個人時，只能自己查看、使用該方案',
            setPublic: '設爲公共',
            publicTip: '設爲公共时，所有用戶都可查看、使用該方案',
            setOrg: '設爲組織',
            orgTip: '設爲組織時，組織內的用戶均能查看、使用該方案',
            setDefault: '設爲默認',
            defaultTip: '下次進入此功能時，默認加載該方案',
            syspreset: '系統預制',
            isDefault: '默認',
            isPublic: '公共',
            isOrg: '組織',
            isPrivate: '個人',
            newScheme: '新增方案',
            mgrScheme: '管理',
            saveas: '另存爲',
            onlyshowcolumns: '僅看展示列',
            schemeList: '方案列表',
            delete: '刪除',
            columns: {
                name: '名稱',
                type: '屬性',
                isDefault: '默認',
                username: '創建人',
                mgr: '管理'
            },
            currentSheme: '當前方案：',
            schemeChanged: {
                title: '方案已更改，是否另存爲新方案？',
                tips: '取消：放棄修改；否：保存當前方案；是：方案另存爲',
                yes: '是',
                no: '否',
                cancel: '取消'
            },
            noDisplyColumnTip: '請設置要顯示的字段'
        },
        'headerGroup': {
            title: '多表頭',
            preview: '預覽效果',
            tips: '在（候選列）中的字段，將不在列表中顯示; 點擊字段名稱，在右側顯示相關操作按鈕。',
            contextMenus: {
                new: '新建表頭',
                newChild: '新建子級表頭',
                setDataField: '分配數據列',
                rename: '重命名',
                top: '置頂',
                up: '上移',
                down: '下移',
                bottom: '置底',
                cancelGroup: '取消分組',
                remove: '移除'
            },
            candidate: '(候選列)',
            setHeader: '設爲表頭',
            ok: '確定',
            cancel: '取消',
            newHeaderPlaceholder: '請輸入名稱',
            newHeaderLabel: '表頭名稱',
        }
    }
};
