// @farris/ui-datagrid-settings/esm5/lib/dbl-click.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Output, EventEmitter } from "@angular/core";
import { fromEvent, race } from 'rxjs';
import { buffer, bufferCount, debounceTime, filter, first, repeat } from 'rxjs/operators';
var FarrisDblClickDirective = /** @class */ (function () {
    function FarrisDblClickDirective(el) {
        this.el = el;
        this.dblClickHandler = new EventEmitter();
        this._click$ = null;
    }
    /**
     * @return {?}
     */
    FarrisDblClickDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var doubleClickDuration = 250;
        /** @type {?} */
        var leftClick$ = fromEvent(this.el.nativeElement, 'click').pipe(filter((/**
         * @param {?} event
         * @return {?}
         */
        function (event) { return event.button === 0; })));
        /** @type {?} */
        var debounce$ = leftClick$.pipe(debounceTime(doubleClickDuration));
        /** @type {?} */
        var clickLimit$ = leftClick$.pipe(bufferCount(2));
        /** @type {?} */
        var bufferGate$ = race(debounce$, clickLimit$)
            .pipe(first(), repeat());
        this._click$ = leftClick$.pipe(buffer(bufferGate$), filter((/**
         * @param {?} clicks
         * @return {?}
         */
        function (clicks) {
            return clicks.length > 1;
        }))).subscribe((/**
         * @param {?} clicks
         * @return {?}
         */
        function (clicks) {
            _this.dblClickHandler.emit(clicks[0]);
        }));
    };
    /**
     * @return {?}
     */
    FarrisDblClickDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._click$) {
            this._click$.unsubscribe();
            this._click$ = null;
        }
    };
    FarrisDblClickDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[f-dbl-click]'
                },] }
    ];
    /** @nocollapse */
    FarrisDblClickDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    FarrisDblClickDirective.propDecorators = {
        options: [{ type: Input, args: ['f-dbl-click',] }],
        dblClickHandler: [{ type: Output }]
    };
    return FarrisDblClickDirective;
}());
export { FarrisDblClickDirective };
if (false) {
    /** @type {?} */
    FarrisDblClickDirective.prototype.options;
    /** @type {?} */
    FarrisDblClickDirective.prototype.dblClickHandler;
    /**
     * @type {?}
     * @private
     */
    FarrisDblClickDirective.prototype._click$;
    /**
     * @type {?}
     * @private
     */
    FarrisDblClickDirective.prototype.el;
}
