// @farris/ui-datagrid-settings/esm5/lib/setting.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function SettingModel() { }
if (false) {
    /** @type {?|undefined} */
    SettingModel.prototype.columnFormat;
    /** @type {?|undefined} */
    SettingModel.prototype.viewColumns;
    /** @type {?|undefined} */
    SettingModel.prototype.groupField;
    /** @type {?|undefined} */
    SettingModel.prototype.sortInfo;
    /** @type {?|undefined} */
    SettingModel.prototype.expandGroupRows;
    /** @type {?|undefined} */
    SettingModel.prototype.multiHeaders;
}
;
/**
 * @record
 */
export function DatagridSchemeModel() { }
if (false) {
    /** @type {?|undefined} */
    DatagridSchemeModel.prototype.id;
    /**
     * 表单中记录Datagrid方案列表的ID，理解为类别，获取Datagrid 方案列表时根据此ID进行查询
     * @type {?}
     */
    DatagridSchemeModel.prototype.belongId;
    /** @type {?|undefined} */
    DatagridSchemeModel.prototype.code;
    /**
     * 是否为默认， 1 默认，0 非默认
     * @type {?|undefined}
     */
    DatagridSchemeModel.prototype.isDefault;
    /**
     * 是否为系统预制
     * @type {?|undefined}
     */
    DatagridSchemeModel.prototype.isSystem;
    /** @type {?|undefined} */
    DatagridSchemeModel.prototype.type;
    /**
     * 方案名称
     * @type {?}
     */
    DatagridSchemeModel.prototype.name;
    /**
     * 方案内容JSON串
     * @type {?|undefined}
     */
    DatagridSchemeModel.prototype.queryConditionString;
    /**
     * 创建人
     * @type {?|undefined}
     */
    DatagridSchemeModel.prototype.userId;
    /** @type {?|undefined} */
    DatagridSchemeModel.prototype.userName;
    /** @type {?|undefined} */
    DatagridSchemeModel.prototype.userCode;
    /** @type {?|undefined} */
    DatagridSchemeModel.prototype.extendId;
    /**
     * 用于存放组织公共配置属于哪些组织，以组织选择组件的返回值为准
     * @type {?|undefined}
     */
    DatagridSchemeModel.prototype.extendInfo;
}
/**
 * @record
 */
export function SetPublicSchemeModel() { }
if (false) {
    /** @type {?} */
    SetPublicSchemeModel.prototype.id;
    /** @type {?} */
    SetPublicSchemeModel.prototype.businessId;
    /** @type {?} */
    SetPublicSchemeModel.prototype.code;
    /** @type {?} */
    SetPublicSchemeModel.prototype.name;
}
/**
 * @record
 */
export function SetOrgAuthSchemeModel() { }
if (false) {
    /** @type {?} */
    SetOrgAuthSchemeModel.prototype.id;
    /** @type {?} */
    SetOrgAuthSchemeModel.prototype.code;
    /** @type {?} */
    SetOrgAuthSchemeModel.prototype.name;
}
/**
 * @record
 */
export function SchemeAuthModel() { }
if (false) {
    /** @type {?} */
    SchemeAuthModel.prototype.success;
    /** @type {?|undefined} */
    SchemeAuthModel.prototype.errorMessage;
    /** @type {?} */
    SchemeAuthModel.prototype.authOperationData;
    /** @type {?} */
    SchemeAuthModel.prototype.authOrganizationData;
    /** @type {?|undefined} */
    SchemeAuthModel.prototype.currentUserId;
}
/**
 * @record
 */
export function BatchScheme() { }
if (false) {
    /** @type {?} */
    BatchScheme.prototype.belongId;
    /** @type {?} */
    BatchScheme.prototype.defaultId;
    /** @type {?} */
    BatchScheme.prototype.deletedIds;
}
/**
 * @record
 */
export function SchemeState() { }
if (false) {
    /** @type {?|undefined} */
    SchemeState.prototype.list;
    /** @type {?|undefined} */
    SchemeState.prototype.current;
}
/**
 * @record
 */
export function DgSchemeState() { }
/**
 * @record
 */
export function ColumnTreeOptions() { }
if (false) {
    /** @type {?|undefined} */
    ColumnTreeOptions.prototype.columns;
    /** @type {?|undefined} */
    ColumnTreeOptions.prototype.treeData;
}
/** @type {?} */
export var IS_MULTI_HEADER_STRING = '-$MULTIHEADER$';
