// @farris/ui-tree/esm2015/lib/show-title.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, Injector } from '@angular/core';
export class ShowTitleDirective {
    /**
     * @param {?} el
     * @param {?} injector
     */
    constructor(el, injector) {
        this.el = el;
        this.injector = injector;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.enableTips) {
            this.el.nativeElement.title = this.el.nativeElement.innerText;
        }
    }
}
ShowTitleDirective.decorators = [
    { type: Directive, args: [{
                selector: '[show-title]'
            },] }
];
/** @nocollapse */
ShowTitleDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Injector }
];
ShowTitleDirective.propDecorators = {
    enableTips: [{ type: Input, args: ['show-title',] }]
};
if (false) {
    /** @type {?} */
    ShowTitleDirective.prototype.enableTips;
    /**
     * @type {?}
     * @private
     */
    ShowTitleDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    ShowTitleDirective.prototype.injector;
}
