// @gsp-wf/wf-process-editor/esm5/lib/views/aif-process-list-help/entity/RtBillProcessListResult.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-bif/common-api';
import { RtShowEntitiesWithCustomization } from './RtShowEntitiesWithCustomization';
var RtBillProcessListResult = /** @class */ (function (_super) {
    tslib_1.__extends(RtBillProcessListResult, _super);
    function RtBillProcessListResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.entities = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RtBillProcessListResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.totalCount = 0;
        this.pageCount = 0;
        this.pageIndex = 0;
        this.pageSize = 0;
        this.entities.length = 0;
    };
    /**
     * @return {?}
     */
    RtBillProcessListResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "totalCount", this.totalCount);
        BpmnModelJsonHelper.SetAttributeValue(obj, "pageCount", this.pageCount);
        BpmnModelJsonHelper.SetAttributeValue(obj, "pageIndex", this.pageIndex);
        BpmnModelJsonHelper.SetAttributeValue(obj, "pageSize", this.pageSize);
        BpmnModelJsonHelper.SetAttributeValue(obj, "entities", this.entities);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RtBillProcessListResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.totalCount = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "totalCount");
        this.pageCount = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "pageCount");
        this.pageIndex = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "pageIndex");
        this.pageSize = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "pageSize");
        this.entities = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "entities", RtShowEntitiesWithCustomization);
    };
    return RtBillProcessListResult;
}(JsonConvertableImpl));
export { RtBillProcessListResult };
if (false) {
    /** @type {?} */
    RtBillProcessListResult.prototype.totalCount;
    /** @type {?} */
    RtBillProcessListResult.prototype.pageCount;
    /** @type {?} */
    RtBillProcessListResult.prototype.pageIndex;
    /** @type {?} */
    RtBillProcessListResult.prototype.pageSize;
    /** @type {?} */
    RtBillProcessListResult.prototype.entities;
}
