// @gsp-wf/wf-process-editor/esm5/lib/views/aif-process-list-help/entity/RtShowEntitiesWithCustomization.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '@edp-bif/common-api';
import { RtShowEntity } from './RtShowEntity';
var RtShowEntitiesWithCustomization = /** @class */ (function (_super) {
    tslib_1.__extends(RtShowEntitiesWithCustomization, _super);
    function RtShowEntitiesWithCustomization() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.customizationEntities = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RtShowEntitiesWithCustomization.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.isEnabled = false;
        this.customizationEntities.length = 0;
    };
    /**
     * @return {?}
     */
    RtShowEntitiesWithCustomization.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "customizationEntities", this.customizationEntities);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RtShowEntitiesWithCustomization.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.customizationEntities = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "customizationEntities", RtShowEntitiesWithCustomization);
    };
    /**
     * @protected
     * @return {?}
     */
    RtShowEntitiesWithCustomization.prototype.getTClass = /**
     * @protected
     * @return {?}
     */
    function () {
        return RtShowEntitiesWithCustomization;
    };
    return RtShowEntitiesWithCustomization;
}(RtShowEntity));
export { RtShowEntitiesWithCustomization };
if (false) {
    /** @type {?} */
    RtShowEntitiesWithCustomization.prototype.isEnabled;
    /** @type {?} */
    RtShowEntitiesWithCustomization.prototype.customizationEntities;
}
