// @farris/ui-verify-detail/esm2015/lib/verify-detail.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { VerifyDetailComponent } from './verify-detail.component';
import { VerifyDetailService } from './verify-detail.service';
import { LocaleModule } from '@farris/ui-locale';
export class VerifyDetailModule {
}
VerifyDetailModule.decorators = [
    { type: NgModule, args: [{
                declarations: [VerifyDetailComponent],
                imports: [
                    CommonModule,
                    FormsModule,
                    LocaleModule
                ],
                providers: [
                    VerifyDetailService
                ],
                exports: [VerifyDetailComponent],
                entryComponents: [VerifyDetailComponent]
            },] }
];
