// @farris/ui-verify-detail/esm2015/lib/verify-detail.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ApplicationRef, Injector, ComponentFactoryResolver } from '@angular/core';
import { VerifyDetailComponent } from './verify-detail.component';
import * as i0 from "@angular/core";
export class VerifyDetailService {
    /**
     * @param {?} cfr
     * @param {?} injector
     * @param {?} appRef
     */
    constructor(cfr, injector, appRef) {
        this.cfr = cfr;
        this.injector = injector;
        this.appRef = appRef;
    }
    /**
     * @param {?} option
     * @return {?}
     */
    createVerify(option) {
        this.clear();
        //if (!this.verifyContainer) {
        /** @type {?} */
        const containerFac = this.cfr.resolveComponentFactory(VerifyDetailComponent);
        /** @type {?} */
        const cmpRef = containerFac.create(this.injector);
        this.appRef.attachView(cmpRef.hostView);
        /** @type {?} */
        let maxHeight;
        if (option.parent) {
            // option.parent.nativeElement.style.position = 'relative';
            maxHeight = option.parent.nativeElement.offsetHeight - 60;
            option.parent.nativeElement.appendChild(cmpRef.location.nativeElement);
        }
        else {
            maxHeight = document.documentElement.clientHeight - 60;
            document.querySelector('body').appendChild(cmpRef.location.nativeElement);
        }
        this.verifyContainer = cmpRef;
        this.verifyContainer.instance.maxHeight = maxHeight;
        this.verifyContainer.instance.showList = option.showList;
        this.verifyContainer.instance.showType = option.showType;
        this.verifyContainer.instance.tabList = option.verifyType;
        this.verifyContainer.instance.validatorList = option.verifyList;
        if (option.hasOwnProperty('verifyResultTipsTmpl') && option['verifyResultTipsTmpl']) {
            this.verifyContainer.instance.resultTipsTmpl = option.verifyResultTipsTmpl;
        }
        // this.verifyContainer.instance.listshowChange.subscribe((event)=>{
        //   console.log(event);
        // });
        // this.verifyContainer.instance.validatorClick.subscribe((event)=>{
        //   console.log(event);
        // });
        return this.verifyContainer.instance;
        //}
    }
    /**
     * @return {?}
     */
    clear() {
        if (this.verifyContainer) {
            /** @type {?} */
            const el = this.verifyContainer.location.nativeElement;
            if (el.parentNode) {
                el.parentNode.removeChild(el);
            }
            this.verifyContainer.destroy();
            this.verifyContainer = undefined;
        }
    }
}
VerifyDetailService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
VerifyDetailService.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: Injector },
    { type: ApplicationRef }
];
/** @nocollapse */ VerifyDetailService.ngInjectableDef = i0.defineInjectable({ factory: function VerifyDetailService_Factory() { return new VerifyDetailService(i0.inject(i0.ComponentFactoryResolver), i0.inject(i0.INJECTOR), i0.inject(i0.ApplicationRef)); }, token: VerifyDetailService, providedIn: "root" });
if (false) {
    /** @type {?} */
    VerifyDetailService.prototype.verifyContainer;
    /**
     * @type {?}
     * @private
     */
    VerifyDetailService.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    VerifyDetailService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    VerifyDetailService.prototype.appRef;
}
