// @edp-aif/common-api/esm5/lib/slim-entity/impl/SlimMappingEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { SlimP2PMetadataEntity } from './SlimP2PMetadataEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * @abstract
 */
var /**
 * @abstract
 */
SlimMappingEntity = /** @class */ (function (_super) {
    tslib_1.__extends(SlimMappingEntity, _super);
    function SlimMappingEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    SlimMappingEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.mappingRuleMetadataId = null;
        this.mappingRuleId = null;
        this.mappingComponentId = null;
        this.mappingClassTypeName = null;
    };
    /**
     * @return {?}
     */
    SlimMappingEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleMetadataId", this.mappingRuleMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleId", this.mappingRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingComponentId", this.mappingComponentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingClassTypeName", this.mappingClassTypeName);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    SlimMappingEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.mappingRuleMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleMetadataId");
        this.mappingRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleId");
        this.mappingComponentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingComponentId");
        this.mappingClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingClassTypeName");
    };
    return SlimMappingEntity;
}(SlimP2PMetadataEntity));
/**
 * @abstract
 */
export { SlimMappingEntity };
if (false) {
    /** @type {?} */
    SlimMappingEntity.prototype.mappingRuleMetadataId;
    /** @type {?} */
    SlimMappingEntity.prototype.mappingComponentId;
    /** @type {?} */
    SlimMappingEntity.prototype.mappingRuleId;
    /** @type {?} */
    SlimMappingEntity.prototype.mappingClassTypeName;
}
