// @edp-aif/common-api/esm5/lib/slim-entity/impl/SlimBizActionEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { SlimMetadataEntity } from './SlimMetadataEntity';
import { BizActionType } from '../../entity/BizActionType';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
var SlimBizActionEntity = /** @class */ (function (_super) {
    tslib_1.__extends(SlimBizActionEntity, _super);
    function SlimBizActionEntity() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.bizActionType = BizActionType.DEFAULT;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    SlimBizActionEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.bizActionType = BizActionType.DEFAULT;
        this.bizActionAgentComponentId = null;
        this.voId = null;
        this.bizActionAgentClassTypeName = null;
    };
    /**
     * @return {?}
     */
    SlimBizActionEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionType", this.bizActionType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionAgentComponentId", this.bizActionAgentComponentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionAgentClassTypeName", this.bizActionAgentClassTypeName);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    SlimBizActionEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.bizActionType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionType");
        this.bizActionAgentComponentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionAgentComponentId");
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
        this.bizActionAgentClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionAgentClassTypeName");
    };
    return SlimBizActionEntity;
}(SlimMetadataEntity));
export { SlimBizActionEntity };
if (false) {
    /** @type {?} */
    SlimBizActionEntity.prototype.bizActionType;
    /** @type {?} */
    SlimBizActionEntity.prototype.voId;
    /** @type {?} */
    SlimBizActionEntity.prototype.bizActionAgentComponentId;
    /** @type {?} */
    SlimBizActionEntity.prototype.bizActionAgentClassTypeName;
}
