// @edp-aif/common-api/esm5/lib/slim-entity/impl/SlimP2PMetadataEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { SlimMetadataEntity } from './SlimMetadataEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * @abstract
 */
var /**
 * @abstract
 */
SlimP2PMetadataEntity = /** @class */ (function (_super) {
    tslib_1.__extends(SlimP2PMetadataEntity, _super);
    function SlimP2PMetadataEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    SlimP2PMetadataEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.sourceBeId = null;
        this.sourceVoId = null;
        this.targetSuCode = null;
        this.targetBeId = null;
        this.targetVoId = null;
        this.targetVoProxyId = null;
    };
    /**
     * @return {?}
     */
    SlimP2PMetadataEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBeId", this.sourceBeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetSuCode", this.targetSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetBeId", this.targetBeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoId", this.targetVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyId", this.targetVoProxyId);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    SlimP2PMetadataEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.sourceBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBeId");
        this.sourceVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
        this.targetSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetSuCode");
        this.targetBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBeId");
        this.targetVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoId");
        this.targetVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyId");
    };
    return SlimP2PMetadataEntity;
}(SlimMetadataEntity));
/**
 * @abstract
 */
export { SlimP2PMetadataEntity };
if (false) {
    /** @type {?} */
    SlimP2PMetadataEntity.prototype.sourceBeId;
    /** @type {?} */
    SlimP2PMetadataEntity.prototype.sourceVoId;
    /** @type {?} */
    SlimP2PMetadataEntity.prototype.targetBeId;
    /** @type {?} */
    SlimP2PMetadataEntity.prototype.targetVoId;
    /** @type {?} */
    SlimP2PMetadataEntity.prototype.targetVoProxyId;
    /** @type {?} */
    SlimP2PMetadataEntity.prototype.targetSuCode;
}
