// @edp-aif/common-api/esm5/lib/slim-entity/impl/SlimCreationRuleEntityEx.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { SlimCreationRuleEntity } from './SlimCreationRuleEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
var SlimCreationRuleEntityEx = /** @class */ (function (_super) {
    tslib_1.__extends(SlimCreationRuleEntityEx, _super);
    function SlimCreationRuleEntityEx(entity) {
        var _this = _super.call(this) || this;
        if (entity != null) {
            _this.id = entity.id;
            _this.description = entity.description;
            _this.mappingComponentId = entity.mappingComponentId;
            _this.mappingRuleMetadataId = entity.mappingRuleMetadataId;
            _this.name = entity.name;
            _this.sourceVoId = entity.sourceVoId;
            _this.sourceBeId = entity.sourceBeId;
            _this.suCode = entity.suCode;
            _this.targetVoId = entity.targetVoId;
            _this.targetBeId = entity.targetBeId;
            _this.targetSuCode = entity.targetSuCode;
            _this.targetVoProxyId = entity.targetVoProxyId;
            _this.formSettingsId = entity.formSettingsId;
            _this.mappingRuleId = entity.mappingRuleId;
        }
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    SlimCreationRuleEntityEx.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.name = null;
        this.sourceVoName = null;
        this.targetVoName = null;
        this.targetVoProxyName = null;
        this.description = null;
    };
    /**
     * @return {?}
     */
    SlimCreationRuleEntityEx.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoName", this.sourceVoName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoName", this.targetVoName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyName", this.targetVoProxyName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "description", this.description);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    SlimCreationRuleEntityEx.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.name = BpmnModelJsonHelper.GetAttributeValue(obj, "name");
        this.sourceVoName = BpmnModelJsonHelper.GetAttributeValue(obj, "sourceVoName");
        this.targetVoName = BpmnModelJsonHelper.GetAttributeValue(obj, "targetVoName");
        this.targetVoProxyName = BpmnModelJsonHelper.GetAttributeValue(obj, "targetVoProxyName");
        this.description = BpmnModelJsonHelper.GetAttributeValue(obj, "description");
    };
    return SlimCreationRuleEntityEx;
}(SlimCreationRuleEntity));
export { SlimCreationRuleEntityEx };
if (false) {
    /** @type {?} */
    SlimCreationRuleEntityEx.prototype.name;
    /** @type {?} */
    SlimCreationRuleEntityEx.prototype.sourceVoName;
    /** @type {?} */
    SlimCreationRuleEntityEx.prototype.targetVoName;
    /** @type {?} */
    SlimCreationRuleEntityEx.prototype.targetVoProxyName;
    /** @type {?} */
    SlimCreationRuleEntityEx.prototype.description;
}
