// @edp-aif/common-api/esm5/lib/slim-entity/impl/SlimAssociationEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { SlimEntity } from './SlimEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
var SlimAssociationEntity = /** @class */ (function (_super) {
    tslib_1.__extends(SlimAssociationEntity, _super);
    function SlimAssociationEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    SlimAssociationEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.srcBeMetadataId = null;
        this.tarBeMetadataId = null;
        this.srcTableName = null;
        this.srcDataId = null;
        this.tarTableName = null;
        this.tarDataId = null;
    };
    /**
     * @return {?}
     */
    SlimAssociationEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcBeMetadataId", this.srcBeMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarBeMetadataId", this.tarBeMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcTableName", this.srcTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcDataId", this.srcDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarTableName", this.tarTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarDataId", this.tarDataId);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    SlimAssociationEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.srcBeMetadataId = BpmnModelJsonHelper.GetAttributeValue(obj, "srcBeMetadataId");
        this.tarBeMetadataId = BpmnModelJsonHelper.GetAttributeValue(obj, "tarBeMetadataId");
        this.srcTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcTableName");
        this.srcDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcDataId");
        this.tarTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarTableName");
        this.tarDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarDataId");
    };
    return SlimAssociationEntity;
}(SlimEntity));
export { SlimAssociationEntity };
if (false) {
    /** @type {?} */
    SlimAssociationEntity.prototype.srcBeMetadataId;
    /** @type {?} */
    SlimAssociationEntity.prototype.tarBeMetadataId;
    /** @type {?} */
    SlimAssociationEntity.prototype.srcTableName;
    /** @type {?} */
    SlimAssociationEntity.prototype.srcDataId;
    /** @type {?} */
    SlimAssociationEntity.prototype.tarTableName;
    /** @type {?} */
    SlimAssociationEntity.prototype.tarDataId;
}
