// @gsp-lcm/metadata-selector/esm2015/metadata/metadata-select/components/metadatapackageselect/metadatapackageselect.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/metadatapackageselect/metadatapackageselect.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, EventEmitter, Output } from '@angular/core';
import { MetadataService } from '../../metadata.service';
import { TreeclassifierComponent } from '../treeclassifier/treeclassifier.component';
import { GSPMetadataService } from '../../../metadata.service';
export class MetadatapackageselectComponent {
    /**
     * @param {?} metadataService
     * @param {?} gspMetaService
     */
    constructor(metadataService, gspMetaService) {
        this.metadataService = metadataService;
        this.gspMetaService = gspMetaService;
        this.closeModalEmitter = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    showModal() {
        // this.isModalShown = true;
    }
    /**
     * @return {?}
     */
    hideModal() {
        // this.isModalShown = false;
        this.closeModalEmitter.emit();
    }
    /**
     * @return {?}
     */
    onHidden() {
        // this.isModalShown = false;
        this.closeModalEmitter.emit();
    }
    /**
     * @param {?} classifier
     * @return {?}
     */
    onSelectedChange(classifier) {
        // this.classifier = classifier;
    }
    /**
     * @return {?}
     */
    ResolvePackage() {
        this.closeModalEmitter.emit();
        if (this.metadataService.packages.length > 0) {
            /** @type {?} */
            let flag = false;
            /** @type {?} */
            const packageName = this.metadataService.packages[0];
            this.gspMetaService.GetDevRootPath().subscribe((/**
             * @param {?} root
             * @return {?}
             */
            root => {
                this.metadataService.rootPath = root;
                this.metadataService.GetMetadataPackageInfo(root + this.metadataService.packRefPath, packageName).subscribe((/**
                 * @param {?} packageInfo
                 * @return {?}
                 */
                packageInfo => {
                    /** @type {?} */
                    const header = packageInfo.header;
                    // 检测依赖项
                    this.metadataService.GetMetadataProjectInfo().subscribe((/**
                     * @param {?} projInfo
                     * @return {?}
                     */
                    projInfo => {
                        // 已有依赖关系
                        /** @type {?} */
                        const refs = projInfo.metadataPackageRefs;
                        refs.forEach((/**
                         * @param {?} item
                         * @return {?}
                         */
                        item => {
                            if (item.name === header.name) {
                                alert('已添加对该元数据包的依赖');
                                flag = true;
                            }
                        }));
                        if (flag === true) {
                            return;
                        }
                        // 更新工程引用
                        this.metadataService.UpdateMetadataProject(header).subscribe((/**
                         * @param {?} data
                         * @return {?}
                         */
                        data => {
                            this.metadataService.treeData = [];
                            this.metadataService.Refresh();
                        }));
                        // 刷新树节点
                        if (this.treeclassifierComponent) {
                            this.treeclassifierComponent.files = this.metadataService.treeData;
                        }
                    }));
                }));
            }));
        }
    }
}
MetadatapackageselectComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'app-metadatapackageselect',
                template: "<!--\u7B2C\u4E8C\u79CD\u5F39\u51FA\u65B9\u6CD5\u7684\u7EC4\u4EF6-->\r\n<!-- <div *ngIf=\"isModalShown\" [config]=\"{ show: true }\" (onHidden)=\"onHidden()\" bsModal #autoShownModal=\"bs-modal\" class=\"modal fade\" tabindex=\"-1\" role=\"dialog\" aria-hidden=\"true\" style='min-height: 800px; min-width: 900px; max-width: none'>\r\n    <div class=\"modal-dialog\" style='max-width:none; width: 615px;margin-top:0'>\r\n        <div class=\"modal-content\"> -->\r\n            <!-- <div class=\"modal-header\">\r\n                <h4 class=\"modal-title pull-left\">\u9009\u62E9\u5143\u6570\u636E\u5305</h4>\r\n                <button type=\"button\" class=\"close pull-right\" aria-label=\"Close\" (click)=\"hideModal()\">\r\n          <span aria-hidden=\"true\">&times;</span>\r\n        </button>\r\n            </div> -->\r\n            <div class=\"modal-body\">\r\n                <div class=\"templateBlock\">\r\n                    <app-metadatapackagelist></app-metadatapackagelist>\r\n                </div>\r\n            </div>\r\n            <div class=\"modal-footer\">\r\n                <button type=\"button\" class=\"btn btn-default btn-primary\" (click)=\"ResolvePackage()\">\u786E\u5B9A</button>\r\n                <button type=\"button\" class=\"btn btn-default\" (click)=\"hideModal()\">\u53D6\u6D88</button>\r\n            </div>\r\n        <!-- </div>\r\n    </div>\r\n</div> -->",
                styles: [".templateBlock{display:flex}.infoBlock{width:100%;min-width:400px}.infoBlock label{width:50px}"]
            }] }
];
/** @nocollapse */
MetadatapackageselectComponent.ctorParameters = () => [
    { type: MetadataService },
    { type: GSPMetadataService }
];
MetadatapackageselectComponent.propDecorators = {
    closeModalEmitter: [{ type: Output }],
    treeclassifierComponent: [{ type: ViewChild, args: [TreeclassifierComponent,] }]
};
if (false) {
    /** @type {?} */
    MetadatapackageselectComponent.prototype.closeModalEmitter;
    /** @type {?} */
    MetadatapackageselectComponent.prototype.treeclassifierComponent;
    /**
     * @type {?}
     * @private
     */
    MetadatapackageselectComponent.prototype.metadataService;
    /**
     * @type {?}
     * @private
     */
    MetadatapackageselectComponent.prototype.gspMetaService;
}
