// @gsp-dip/data-imp-exp/esm5/lib/service/Entity/CommonParams.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CommonParams = /** @class */ (function () {
    function CommonParams() {
        this.ruleId = '';
        this.keys = [];
        this.option = {};
        this.fileName = '';
        this.frameContext = null;
        this.repository = null;
        this.befRestService = null;
        this.suffix = '';
        this.loadingParam = {};
        this.i18nResource = {};
        this.importParam = {};
        this.mainObjId = '';
        this.su = '';
        this.importResult = {};
        this.securityLevelName = '';
    }
    return CommonParams;
}());
export { CommonParams };
if (false) {
    /** @type {?} */
    CommonParams.prototype.ruleId;
    /** @type {?} */
    CommonParams.prototype.keys;
    /** @type {?} */
    CommonParams.prototype.option;
    /** @type {?} */
    CommonParams.prototype.fileName;
    /** @type {?} */
    CommonParams.prototype.frameContext;
    /** @type {?} */
    CommonParams.prototype.repository;
    /** @type {?} */
    CommonParams.prototype.befRestService;
    /** @type {?} */
    CommonParams.prototype.suffix;
    /** @type {?} */
    CommonParams.prototype.loadingParam;
    /** @type {?} */
    CommonParams.prototype.i18nResource;
    /** @type {?} */
    CommonParams.prototype.importParam;
    /** @type {?} */
    CommonParams.prototype.mainObjId;
    /** @type {?} */
    CommonParams.prototype.su;
    /** @type {?} */
    CommonParams.prototype.importResult;
    /** @type {?} */
    CommonParams.prototype.securityLevelName;
    /** @type {?} */
    CommonParams.prototype.methodType;
}
