// @gspwidget/widget-core/esm2015/lib/title-bar/title-bar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ContentChild, EventEmitter, Input, Output, TemplateRef } from '@angular/core';
import { ArrUtil, deepClone, Indicator, objectLiteralEqual, RuntimeAggOption } from '@gspwidget/util';
import { ifFsDateRange, isFsFlat, JumpTypes, WidgetDevkitService } from '@gspwidget/widget-devkit';
import { TranslateService } from '@ngx-translate/core';
export class TitleBarComponent {
    /**
     * @param {?} devkitSvc
     * @param {?} translate
     */
    constructor(devkitSvc, translate) {
        this.devkitSvc = devkitSvc;
        this.translate = translate;
        this.icon = '';
        /**
         * 标题
         */
        this.showTitle = true;
        this.title = '';
        /**
         * 数字
         */
        this.showNumber = false;
        this.number = 0;
        /**
         * 设置按钮
         */
        this.showSetting = false;
        this.clickSetting = new EventEmitter();
        /**
         * 刷新按钮
         */
        this.showRefresh = false;
        this.clickRefresh = new EventEmitter();
        /**
         * 维度，用于运行时聚合
         */
        this.dims = [];
        /**
         * 是否显示运行时聚合
         */
        this.showRuntimeAgg = false;
        this.filterChange = new EventEmitter();
        this.runtimeAggChange = new EventEmitter();
        this.init = new EventEmitter();
        this.JumpTypes = JumpTypes;
        /**
         * 聚合
         */
        this.editingRuntimeAgg = new RuntimeAggOption();
        this.runtimeAggDropdownOpen = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.init.emit(this);
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.filterConfigs) {
            this.outerFC = null;
            this.innerFCs = null;
            /** @type {?} */
            const notFlatFCs = this.filterConfigs.filter((/**
             * @param {?} fc
             * @return {?}
             */
            fc => !isFsFlat(fc.fs)));
            if (notFlatFCs && notFlatFCs[0])
                this.outerFC = notFlatFCs[0];
            if (notFlatFCs && notFlatFCs.length > 1)
                this.innerFCs = notFlatFCs.slice(1);
            this.flatFCs = this.filterConfigs.filter((/**
             * @param {?} fc
             * @return {?}
             */
            fc => isFsFlat(fc.fs)));
        }
    }
    /**
     * 用于外部手动设置 runtime agg。因为每次都需要 deep clone，所以封装为手动调用的函数，而非模板绑定，以减少次数。
     * @param {?} runtimeAgg
     * @return {?}
     */
    setRuntimeAgg(runtimeAgg) {
        if (!runtimeAgg)
            return;
        this.editingRuntimeAgg = deepClone(runtimeAgg);
        this.confirmedRuntimeAgg = deepClone(runtimeAgg);
    }
    /**
     * @return {?}
     */
    handleClickJump() {
        this.devkitSvc.jumpWithoutData(this.jumpConfig);
    }
    /**
     * 把 filter 发来的结果值，加上 dpId 一起包装为 dpValues 发出去
     * @param {?} value
     * @param {?} fc
     * @return {?}
     */
    handleFilterSingleChange(value, fc) {
        // 1. 普通：发出对应 1 个 dp 的单个值
        this.filterChange.emit([{
                value,
                dpId: fc.dpId,
            }]);
    }
    /**
     * @param {?} values
     * @param {?} fc
     * @return {?}
     */
    handleFilterMultiChange(values, fc) {
        /** @type {?} */
        let dpValues
        // 2. 日期范围：发出对应 2 个 dp 的 2 个值
        ;
        // 2. 日期范围：发出对应 2 个 dp 的 2 个值
        if (ifFsDateRange(fc.fs)) {
            dpValues = [
                { value: values[0], dpId: fc.dpId },
                { value: values[1], dpId: fc.dpId2 },
            ];
        }
        else {
            // 3. 枚举多选：发出对应 1 个 dp 的多个值
            dpValues = [{
                    // TODO: 多值的连接方式是否可配？
                    value: values.join(','), dpId: fc.dpId
                }];
        }
        this.filterChange.emit(dpValues);
    }
    /**
     * 聚合
     * @return {?}
     */
    handleAddInd() {
        this.editingRuntimeAgg.indicators.push(new Indicator());
    }
    /**
     * @param {?} ind
     * @return {?}
     */
    handleRemoveInd(ind) {
        ArrUtil.Remove(this.editingRuntimeAgg.indicators, ind);
    }
    /**
     * @return {?}
     */
    handleConfirmAgg() {
        this.runtimeAggDropdownOpen = false;
        if (objectLiteralEqual(this.editingRuntimeAgg, this.confirmedRuntimeAgg))
            return;
        this.confirmedRuntimeAgg = deepClone(this.editingRuntimeAgg);
        this.runtimeAggChange.emit(this.confirmedRuntimeAgg);
    }
}
TitleBarComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-title-bar',
                template: "<!-- --{{filterSchemas | json}}-- -->\n<!-- TODO: \u6240\u6709\u5176\u4ED6 filter \u6536\u8D77\u5230 dropdown \u91CC -->\n<!-- <div class=\"d-flex\">\n  <filter\n    *ngFor=\"let fc of filterConfigs\"\n    [filterSchema]=\"fc.fs\"\n    (singleValueChange)=\"handleFilterSingleChange($event, fc)\"\n    (multiValueChange)=\"handleFilterMultiChange($event, fc)\"\n    ></filter>\n</div> -->\n<div class=\"title-bar\">\n  <div class=\"d-flex\">\n    <!-- left -->\n    <div class=\"left align-items-center\">\n      <i *ngIf=\"icon!=''\" class=\"material-icons\">{{icon}}</i>\n      <div *ngIf=\"showTitle\" class=\"title\">{{title}}</div>\n      <div *ngIf=\"showNumber\" class=\"number\">{{number}}</div>\n      <ng-container *ngTemplateOutlet=\"leftTemplate\"></ng-container>\n    </div>\n    <!-- right -->\n    <div class=\"right\">\n      <ng-container *ngTemplateOutlet=\"rightTemplate\"></ng-container>\n      <filter\n        *ngIf=\"outerFC\"\n        [filterSchema]=\"outerFC.fs\"\n        (singleValueChange)=\"handleFilterSingleChange($event, outerFC)\"\n        (multiValueChange)=\"handleFilterMultiChange($event, outerFC)\"\n        ></filter>\n      <div *ngIf=\"innerFCs\"\n        class=\"tool-wrap d-flex\">\n        <nz-dropdown nzTrigger=\"click\">\n          <i nz-dropdown class=\"material-icons\">more_horiz</i>\n          <ul nz-menu class=\"p-2\">\n            <li\n              *ngFor=\"let fc of innerFCs\"\n              class=\"mt-1 d-flex align-items-center\">\n              <span class=\"text-truncate\" style=\"width: 50px\">{{fc?.fs?.i18nName[translate.currentLang]}}</span>\n              <filter\n                [filterSchema]=\"fc.fs\"\n                (singleValueChange)=\"handleFilterSingleChange($event, fc)\"\n                (multiValueChange)=\"handleFilterMultiChange($event, fc)\"\n                ></filter>\n            </li>\n          </ul>\n        </nz-dropdown>\n      </div>\n      <!-- \u805A\u5408\u914D\u7F6E -->\n      <div *ngIf=\"showRuntimeAgg\"\n        class=\"tool-wrap d-flex\">\n        <nz-dropdown nzTrigger=\"click\" [(nzVisible)]=\"runtimeAggDropdownOpen\">\n          <i nz-dropdown class=\"material-icons\">show_chart</i>\n          <div nz-menu class=\"p-2\" style=\"width:200px;box-sizing:content-box;\">\n            <h5>\u805A\u5408\u914D\u7F6E</h5>\n            <div class=\"tip p-2\">\n              \u805A\u5408\u914D\u7F6E\u4EC5\u5BF9 QDP \u6570\u636E\u96C6\u751F\u6548\u3002\n            </div>\n            <div class=\"sm-title\">\u7EF4\u5EA6</div>\n            <nz-select\n              [(ngModel)]=\"editingRuntimeAgg.dimensions[0]\"\n              nzPlaceHolder=\"\u9009\u62E9\u7EF4\u5EA6\u5B57\u6BB5\"\n              style=\"width:200px\">\n              <nz-option\n                *ngFor=\"let dim of dims\"\n                [nzLabel]=\"dim.label || dim.name\"\n                [nzValue]=\"dim.name\"></nz-option>\n            </nz-select>\n            <div class=\"sm-title mt-2\">\u6C47\u603B\u65B9\u5F0F\u53CA\u6307\u6807</div>\n            <ul>\n              <li\n                class=\"mb-1 d-flex align-items-center\"\n                *ngFor=\"let ind of editingRuntimeAgg.indicators\">\n                <nz-input-group nzCompact>\n                  <nz-select [(ngModel)]=\"ind.aggFn\" style=\"width:70px\">\n                    <nz-option nzLabel=\"\u5408\u8BA1\" [nzValue]=\"1\"></nz-option>\n                    <nz-option nzLabel=\"\u6700\u5927\u503C\" [nzValue]=\"2\"></nz-option>\n                    <nz-option nzLabel=\"\u6700\u5C0F\u503C\" [nzValue]=\"3\"></nz-option>\n                    <nz-option nzLabel=\"\u8BA1\u6570\" [nzValue]=\"4\"></nz-option>\n                    <nz-option nzLabel=\"\u5E73\u5747\u503C\" [nzValue]=\"5\"></nz-option>\n                  </nz-select>\n                  <nz-select\n                    nzPlaceHolder=\"\u9009\u62E9\u6307\u6807\u5B57\u6BB5\"\n                    [(ngModel)]=\"ind.dimName\"\n                    style=\"width:110px\">\n                    <nz-option\n                      *ngFor=\"let dim of dims\"\n                      [nzLabel]=\"dim.label || dim.name\"\n                      [nzValue]=\"dim.name\"></nz-option>\n                  </nz-select>\n                </nz-input-group>\n                <i\n                  class=\"material-icons cursor-pointer fs-15 text-gray-200 hover:text-red-600 transition-colors\"\n                  (click)=\"handleRemoveInd(ind)\"\n                  >remove_circle</i>\n              </li>\n            </ul>\n\n            <div class=\"text-right mt-2\">\n              <button class=\"btn btn-secondary mr-1\" (click)=\"handleAddInd()\">\u589E\u52A0\u6307\u6807</button>\n              <button class=\"btn btn-primary\" (click)=\"handleConfirmAgg()\">{{'confirm' | translate}}</button>\n            </div>\n          </div>\n\n          <!-- <ul nz-menu class=\"p-2\">\n            <li\n              *ngFor=\"let fc of innerFCs\"\n              class=\"mt-1 d-flex align-items-center\">\n              <span class=\"text-truncate\" style=\"width: 50px\">{{fc?.fs?.i18nName[translate.currentLang]}}</span>\n              <filter\n                [filterSchema]=\"fc.fs\"\n                (singleValueChange)=\"handleFilterSingleChange($event, fc)\"\n                (multiValueChange)=\"handleFilterMultiChange($event, fc)\"\n                ></filter>\n            </li>\n          </ul> -->\n        </nz-dropdown>\n      </div>\n      <div\n        class=\"tool-wrap d-flex\"\n        [title]=\"(settingTitle || '') | translate\"\n        *ngIf=\"showSetting\">\n        <img\n          class=\"setting\"\n          (click)=\"clickSetting.emit()\"\n          src=\"assets/img/setting.svg\"\n          [alt]=\"settingTitle\">\n      </div>\n      <div\n        class=\"tool-wrap d-flex\"\n        *ngIf=\"showRefresh\"\n        (click)=\"clickRefresh.emit()\">\n        <i class=\"icon-refresh material-icons\" style=\"font-size: 20px;\">refresh</i>\n      </div>\n      <!-- -{{jumpConfig.jumpType}}-{{jumpConfig|json}}- -->\n      <div\n        *ngIf=\"jumpConfig && jumpConfig.jumpType!==JumpTypes.None\"\n        (click)=\"handleClickJump()\"\n        class=\"tool-wrap d-flex\">\n        <i class=\"material-icons\">chevron_right</i>\n      </div>\n    </div>\n  </div>\n  <ul>\n    <li *ngFor=\"let flatFC of flatFCs\" class=\"mt-1\">\n      <filter\n        [filterSchema]=\"flatFC.fs\"\n        (singleValueChange)=\"handleFilterSingleChange($event, flatFC)\"\n        (multiValueChange)=\"handleFilterMultiChange($event, flatFC)\"\n        ></filter>\n    </li>\n  </ul>\n</div>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: ["@charset \"UTF-8\";.title-bar{padding:12px 20px 10px}.title-bar .left{flex:1 1 auto;display:flex}.title-bar .left .material-icons{font-size:18px;margin-right:8px;color:#3794ff;padding-top:1px}.title-bar .left .title{font-weight:700;font-size:15px}.title-bar .left .number{font-size:15px;margin-left:10px}.title-bar .right{flex:1 1 auto;display:flex;justify-content:flex-end;align-items:center}.title-bar .right .widget-select{margin-left:5px}.title-bar .right .setting{width:16px;height:16px;transition:.3s;opacity:1!important}.title-bar .right .setting:hover{transform:rotate(60deg)}.title-bar .right .tool-wrap{width:20px;margin-left:10px;font-size:14px;color:#000;transition:.3s;cursor:pointer;opacity:.2}.title-bar .right .tool-wrap:hover{opacity:.4}"]
            }] }
];
/** @nocollapse */
TitleBarComponent.ctorParameters = () => [
    { type: WidgetDevkitService },
    { type: TranslateService }
];
TitleBarComponent.propDecorators = {
    icon: [{ type: Input }],
    showTitle: [{ type: Input }],
    title: [{ type: Input }],
    showNumber: [{ type: Input }],
    number: [{ type: Input }],
    showSetting: [{ type: Input }],
    settingTitle: [{ type: Input }],
    clickSetting: [{ type: Output }],
    showRefresh: [{ type: Input }],
    clickRefresh: [{ type: Output }],
    jumpConfig: [{ type: Input }],
    jumpProperty: [{ type: Input }],
    data: [{ type: Input }],
    filterConfigs: [{ type: Input }],
    dims: [{ type: Input }],
    showRuntimeAgg: [{ type: Input }],
    filterChange: [{ type: Output }],
    runtimeAggChange: [{ type: Output }],
    init: [{ type: Output }],
    leftTemplate: [{ type: ContentChild, args: ['leftTemplate',] }],
    rightTemplate: [{ type: ContentChild, args: ['rightTemplate',] }]
};
if (false) {
    /** @type {?} */
    TitleBarComponent.prototype.icon;
    /**
     * 标题
     * @type {?}
     */
    TitleBarComponent.prototype.showTitle;
    /** @type {?} */
    TitleBarComponent.prototype.title;
    /**
     * 数字
     * @type {?}
     */
    TitleBarComponent.prototype.showNumber;
    /** @type {?} */
    TitleBarComponent.prototype.number;
    /**
     * 设置按钮
     * @type {?}
     */
    TitleBarComponent.prototype.showSetting;
    /** @type {?} */
    TitleBarComponent.prototype.settingTitle;
    /** @type {?} */
    TitleBarComponent.prototype.clickSetting;
    /**
     * 刷新按钮
     * @type {?}
     */
    TitleBarComponent.prototype.showRefresh;
    /** @type {?} */
    TitleBarComponent.prototype.clickRefresh;
    /**
     * 是否启用跳转，开启后元部件配置中可配置全局联查
     * @type {?}
     */
    TitleBarComponent.prototype.jumpConfig;
    /**
     * @deprecated 应改用 enableJump。全局联查配置
     * @type {?}
     */
    TitleBarComponent.prototype.jumpProperty;
    /**
     * @deprecated 应改用 enableJump。为了跳转时能取到数据的第一行，用来拼接跳转路由参数。
     * @type {?}
     */
    TitleBarComponent.prototype.data;
    /**
     * 筛选
     * @type {?}
     */
    TitleBarComponent.prototype.filterConfigs;
    /**
     * 维度，用于运行时聚合
     * @type {?}
     */
    TitleBarComponent.prototype.dims;
    /**
     * 是否显示运行时聚合
     * @type {?}
     */
    TitleBarComponent.prototype.showRuntimeAgg;
    /** @type {?} */
    TitleBarComponent.prototype.filterChange;
    /** @type {?} */
    TitleBarComponent.prototype.runtimeAggChange;
    /** @type {?} */
    TitleBarComponent.prototype.init;
    /**
     * 传入模板
     * @type {?}
     */
    TitleBarComponent.prototype.leftTemplate;
    /** @type {?} */
    TitleBarComponent.prototype.rightTemplate;
    /** @type {?} */
    TitleBarComponent.prototype.JumpTypes;
    /**
     * 非 flat 的 fcs
     * @type {?}
     */
    TitleBarComponent.prototype.outerFC;
    /** @type {?} */
    TitleBarComponent.prototype.innerFCs;
    /**
     * flat 的 fc
     * @type {?}
     */
    TitleBarComponent.prototype.flatFCs;
    /**
     * 聚合
     * @type {?}
     */
    TitleBarComponent.prototype.editingRuntimeAgg;
    /** @type {?} */
    TitleBarComponent.prototype.confirmedRuntimeAgg;
    /** @type {?} */
    TitleBarComponent.prototype.runtimeAggDropdownOpen;
    /**
     * @type {?}
     * @private
     */
    TitleBarComponent.prototype.devkitSvc;
    /** @type {?} */
    TitleBarComponent.prototype.translate;
}
