// @farris/external-integration/esm2015/lib/directives/ext-integration-data-mapping.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ext-integration-data-mapping.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, Input, Optional, Output } from "@angular/core";
import { ViewModel } from "@farris/devkit";
import { ExternalIntegrationComponent } from "../../lib/component/external-integration/external-integration.component";
export class ExtIntegrationDataMappingDirective {
    /**
     * @param {?} vm
     * @param {?} extIntegrationComponent
     */
    constructor(vm, extIntegrationComponent) {
        this.vm = vm;
        this.extIntegrationComponent = extIntegrationComponent;
        this.showModal = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.extIntegrationComponent.selectSapData.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            this.mappingData(data, this.myField);
        }));
    }
    /**
     * 字段映射
     * @param {?} sapData sap返回信息
     * @param {?} mappingField 映射关系
     * @return {?}
     */
    mappingData(sapData, mappingField) {
        /** @type {?} */
        let helpFields = Object.keys(mappingField);
        helpFields.forEach((/**
         * @param {?} fieldName
         * @return {?}
         */
        (fieldName) => {
            /** @type {?} */
            let pathArr = this.getBindingPathArray();
            // 获取字段值
            /** @type {?} */
            let val;
            if (sapData) {
                val = sapData[fieldName];
            }
            this.vm.bindingData.setValue(pathArr.concat(mappingField[fieldName]), val, true, true);
        }));
    }
    /**
     * @private
     * @return {?}
     */
    getBindingPathArray() {
        /** @type {?} */
        const path = this.vm.bindingPath;
        if (path) {
            return path.split("/").filter((/**
             * @param {?} n
             * @return {?}
             */
            (n) => n !== ""));
        }
        return [];
    }
}
ExtIntegrationDataMappingDirective.decorators = [
    { type: Directive, args: [{
                selector: "[extIntegration-dataMapping]",
            },] }
];
/** @nocollapse */
ExtIntegrationDataMappingDirective.ctorParameters = () => [
    { type: ViewModel, decorators: [{ type: Optional }] },
    { type: ExternalIntegrationComponent }
];
ExtIntegrationDataMappingDirective.propDecorators = {
    myField: [{ type: Input, args: ["mapFields",] }],
    showModal: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    ExtIntegrationDataMappingDirective.prototype.myField;
    /** @type {?} */
    ExtIntegrationDataMappingDirective.prototype.showModal;
    /**
     * @type {?}
     * @private
     */
    ExtIntegrationDataMappingDirective.prototype.vm;
    /**
     * @type {?}
     * @private
     */
    ExtIntegrationDataMappingDirective.prototype.extIntegrationComponent;
}
