// @farris/ui-field-group/esm5/lib/directive/tab-index.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ElementRef, Injector, Renderer2, Self, Optional } from '@angular/core';
import { EditorComponent } from '@farris/ui-editor';
/**
 * form 表单中输入组件的tab-index
 * Date: 2021-03-11
 * Creator: Lucas GOGO
 */
import { Directive, Input } from '@angular/core';
var TabIndexDirective = /** @class */ (function () {
    function TabIndexDirective(elRef, inject, render, editorRef) {
        this.elRef = elRef;
        this.inject = inject;
        this.render = render;
        this.editorRef = editorRef;
        this.input = 'INPUT';
        this.textarea = 'TEXTAREA';
        this.inputGroup = 'INPUT-GROUP';
        this.lookup = 'FARRIS-LOOKUP-GRID';
        this.comboLookup = 'FARRIS-COMBO-LOOKUP';
        this.comboList = 'FARRIS-COMBO-LIST';
        this.datepicker = 'FARRIS-DATEPICKER';
        this.timepicker = 'FARRIS-TIME-PICKER';
        this.numberSpinner = 'FARRIS-NUMBER-SPINNER';
        this.checkGroup = 'FARRIS-CHECKBOXGROUP';
        this.radioGroup = 'FARRIS-RADIOGROUP';
        this.switchCase = 'FARRIS-SWITCH';
        this.htmlEditor = 'FARRIS-EDITOR';
        this.colorPicker = 'COLOR-PICKER';
        this.langText = 'LANGUAGE-TEXTBOX';
        this.inputs = [this.input, this.textarea];
        this.farrisInputs = [this.inputGroup, this.lookup, this.comboList, this.colorPicker, this.langText,
            this.comboLookup, this.datepicker, this.timepicker, this.numberSpinner];
    }
    /**
     * @return {?}
     */
    TabIndexDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    TabIndexDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.updateTabIndex();
        }));
    };
    /**
     * @private
     * @return {?}
     */
    TabIndexDirective.prototype.updateTabIndex = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var el = this.elRef.nativeElement;
        if (this.tabIndex === '' || this.tabIndex === undefined || this.tabIndex === null) {
            this.tabIndex = el.tabIndex !== undefined ? el.tabIndex : -1;
        }
        /** @type {?} */
        var tagName = el.tagName.toUpperCase();
        /** @type {?} */
        var inputEle = null;
        if (this.inputs.indexOf(tagName) > -1) {
            inputEle = el;
        }
        else {
            if (this.farrisInputs.indexOf(tagName) > -1) {
                inputEle = el.querySelector('input');
            }
            if (tagName === this.htmlEditor) {
                if (this.editorRef) {
                    this.editorRef.onInit.subscribe((/**
                     * @return {?}
                     */
                    function () {
                        inputEle = el.querySelector('iframe');
                        el.removeAttribute('tabindex');
                        inputEle.tabIndex = _this.tabIndex;
                    }));
                }
            }
        }
        if (inputEle) {
            el.removeAttribute('tabindex');
            inputEle.tabIndex = this.tabIndex;
        }
    };
    TabIndexDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[tab-index]',
                },] }
    ];
    /** @nocollapse */
    TabIndexDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Injector },
        { type: Renderer2 },
        { type: EditorComponent, decorators: [{ type: Optional }, { type: Self }] }
    ]; };
    TabIndexDirective.propDecorators = {
        tabIndex: [{ type: Input, args: ['tab-index',] }]
    };
    return TabIndexDirective;
}());
export { TabIndexDirective };
if (false) {
    /** @type {?} */
    TabIndexDirective.prototype.tabIndex;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.input;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.textarea;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.inputGroup;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.lookup;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.comboLookup;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.comboList;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.datepicker;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.timepicker;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.numberSpinner;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.checkGroup;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.radioGroup;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.switchCase;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.htmlEditor;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.colorPicker;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.langText;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.inputs;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.farrisInputs;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.elRef;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.inject;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.render;
    /**
     * @type {?}
     * @private
     */
    TabIndexDirective.prototype.editorRef;
}
