// @farris/expression-engine/esm2015/lib/expression/expression.js
import { Compiler } from "../compiler/index";
import { Evaler } from "../evaler/index";
import { DefaultFunctions } from "../function/index";
export class Expression {
    constructor(expr, context) {
        this.expr = expr;
        this.context = context;
    }
    eval(context) {
        context = context || this.context;
        const args = context && context.contexts && context.contexts.arguments || {};
        if (this.factory) {
            const ctx = this.buildContext(context);
            const result = this.factory(ctx.contexts)(args);
            return result;
        }
        else {
            const evaler = new Evaler();
            const factory = evaler.eval(this.expr, context);
            const result = factory(...Object.values(args));
            return result;
        }
    }
    compile(context) {
        context = context || this.context;
        const compiler = new Compiler();
        this.factory = compiler.compile(this.expr, context);
        return this;
    }
    buildContext(context) {
        const contexts = context && context.contexts || {};
        const defaultFunction = new DefaultFunctions(context);
        contexts['DefaultFunction'] = defaultFunction;
        return context;
    }
}
