// @gsp-wf/wf-process-editor/esm5/lib/views/performer/views/activity-executor-performer/activity-executor-performer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, NgZone, ViewChild, ElementRef, HostBinding, ChangeDetectorRef, Optional, Inject, LOCALE_ID } from '@angular/core';
import { DialogComponent } from "@farris/ui-dialog";
import { DatagridComponent } from "@farris/ui-datagrid";
import { ExprEditMode, ExpressionService, InjectedType } from "@gsp-svc/expression";
import { ProcessDesignerUIState } from "../../../../app/process-designer.uistate";
import { ActivityExecutorParticipant } from "./entities/activity-executor.participant";
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../../../i18n/pipe/wf-process-editor-locale.pipe';
var ActivityExecutorPerformerComponent = /** @class */ (function () {
    function ActivityExecutorPerformerComponent(ngZone, designerUIState, expr, cdr, localeId, resources, localePipe) {
        this.ngZone = ngZone;
        this.designerUIState = designerUIState;
        this.expr = expr;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.cls = 'd-flex flex-fill h-100';
        this.showClearButton = false;
        this.activityExecutors = [];
        this.cols = [
            { title: this.localePipe.transform('component.activityExecutor.name'), field: 'Name', width: 60 },
            { title: 'ID', field: 'Id', width: 90 }
        ];
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
    }
    /**
     * @return {?}
     */
    ActivityExecutorPerformerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    ActivityExecutorPerformerComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            _this.inputGroup.nativeElement.addEventListener('mouseenter', _this.onMouseEnter.bind(_this));
            _this.inputGroup.nativeElement.addEventListener('mouseleave', _this.onMouseLeave.bind(_this));
        }));
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ActivityExecutorPerformerComponent.prototype.onMouseEnter = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.showClearButton = true;
        if (this.showClearButton) {
            this.toggleClearIcon(event, true);
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ActivityExecutorPerformerComponent.prototype.onMouseLeave = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.showClearButton = false;
        this.toggleClearIcon(event, false);
    };
    /**
     * @private
     * @param {?} event
     * @param {?=} isShow
     * @return {?}
     */
    ActivityExecutorPerformerComponent.prototype.toggleClearIcon = /**
     * @private
     * @param {?} event
     * @param {?=} isShow
     * @return {?}
     */
    function (event, isShow) {
        if (isShow === void 0) { isShow = false; }
        /** @type {?} */
        var str = isShow ? '' : 'none';
        /** @type {?} */
        var clearIcon = event.target.querySelector('.input-group-clear');
        if (clearIcon) {
            clearIcon.style.display = str;
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    ActivityExecutorPerformerComponent.prototype.clear = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.activityExecutors = [];
    };
    /**
     * @param {?} data
     * @return {?}
     */
    ActivityExecutorPerformerComponent.prototype.selectUsers = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var e_1, _a;
        // 已办用户
        /** @type {?} */
        var selectedUser = [];
        if (this.activityExecutors && this.activityExecutors.length > 0) {
            try {
                for (var _b = tslib_1.__values(this.activityExecutors), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var handler = _c.value;
                    /** @type {?} */
                    var data_1 = {
                        data: {
                            Id: handler.activityDefId,
                            Name: handler.activityDefName
                        },
                        id: handler.activityDefId
                    };
                    selectedUser.push(data_1);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        this.allActivityTable.setCurrentPagerData(selectedUser);
        this.allActivityDialog.show();
    };
    /**
     * @return {?}
     */
    ActivityExecutorPerformerComponent.prototype.editConditon = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var e_2, _a, e_3, _b, e_4, _c, e_5, _d;
        // this.expr.clearContext();
        if (this.designerUIState.schemas.length > 0) {
            try {
                for (var _e = tslib_1.__values(this.designerUIState.schemas), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var schema = _f.value;
                    this.expr.addSchema('Schema', schema);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_a = _e.return)) _a.call(_e);
                }
                finally { if (e_2) throw e_2.error; }
            }
            if (this.designerUIState.schemas[0].entityTypes && this.designerUIState.schemas[0].entityTypes.length > 0) {
                /** @type {?} */
                var entityType = this.designerUIState.schemas[0].entityTypes[0];
                this.expr.addInjectedEntity(entityType.name, 'Schema', entityType.name, InjectedType.EntityType);
            }
        }
        /** @type {?} */
        var context = new Array();
        context = this.designerUIState.addContext(context);
        if (this.designerUIState.subActivityContext.length > 0) {
            var _loop_1 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _g = tslib_1.__values(this.designerUIState.subActivityContext), _h = _g.next(); !_h.done; _h = _g.next()) {
                    var variable = _h.value;
                    _loop_1(variable);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_h && !_h.done && (_b = _g.return)) _b.call(_g);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        if (this.designerUIState.activityComponentContext.length > 0) {
            var _loop_2 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _j = tslib_1.__values(this.designerUIState.activityComponentContext), _k = _j.next(); !_k.done; _k = _j.next()) {
                    var variable = _k.value;
                    _loop_2(variable);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_k && !_k.done && (_c = _j.return)) _c.call(_j);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
        if (context.length > 0) {
            try {
                for (var context_1 = tslib_1.__values(context), context_1_1 = context_1.next(); !context_1_1.done; context_1_1 = context_1.next()) {
                    var v = context_1_1.value;
                    this.expr.addContext(v);
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (context_1_1 && !context_1_1.done && (_d = context_1.return)) _d.call(context_1);
                }
                finally { if (e_5) throw e_5.error; }
            }
        }
        this.expr.setEditPattern(ExprEditMode.Conditional); // 表达式展现模式：默认简易版
        this.expr.addExpressionText(this.condition);
        this.expr.buildExpression().then((/**
         * @param {?} expressioninfo
         * @return {?}
         */
        function (expressioninfo) {
            _this.condition = expressioninfo;
            //点击确定后，去除表达式里面的新增的上下文变量
            if (context.length > 0) {
                _this.expr.exprEntity.contextEntities = _this.expr.exprEntity.contextEntities.filter((/**
                 * @param {?} x
                 * @return {?}
                 */
                function (x) { return !context.some((/**
                 * @param {?} y
                 * @return {?}
                 */
                function (y) { return y.key === x.key; })); }));
            }
        }));
    };
    /**
     * @return {?}
     */
    ActivityExecutorPerformerComponent.prototype.activitiesSelect = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var selectedActivities = this.allActivityTable.checkeds;
        this.activityExecutors = selectedActivities.map((/**
         * @param {?} activity
         * @return {?}
         */
        function (activity) {
            /** @type {?} */
            var handler = new ActivityExecutorParticipant(activity.data.Id, activity.data.Name, 'wf.ActivityExecutorParticipantExpression');
            handler.performerName = '<' + activity.data.Name + _this.localePipe.transform('component.activityExecutor.eventHandler');
            handler.activityDefId = activity.data.Id;
            handler.activityDefName = activity.data.Name;
            handler.condition = _this.condition;
            return handler;
        }));
        this.allActivityDialog.close();
    };
    /**
     * @return {?}
     */
    ActivityExecutorPerformerComponent.prototype.getPerformers = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.activityExecutors.map((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.condition = _this.condition; }));
        return this.activityExecutors;
    };
    ActivityExecutorPerformerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-activity-handler-performer',
                    template: "<div class=\"d-flex flex-fill farris-overflow-y-auto p-md-2\" style=\"background: #FFF;\">\r\n  <div class=\"d-flex flex-fill farris-overflow-y-auto mx-3 my-2\" style=\"border-radius:16px;border: 1.5px solid #F1F3F7;\">\r\n  <form style=\"width: 100%;\" >\r\n    <div class=\"d-flex m-2 flex-column w-100\">\r\n      <div class=\"d-flex flex-column\">\r\n        <div class=\"farris-form form-inline  farris-form-inline mr-5 my-2\">\r\n          <div class=\"col-12 farris-group-auto my-4\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-info text-danger\">*</span>\r\n                  <span class=\"farris-label-text\">{{'component.activityExecutor.nodeName'|wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <div #inputGroup class=\"input-group\">\r\n                    <input type=\"text\" class=\"form-control\" [value]=\"activityExecutors | activityExecutor \" name=\"users\" [readOnly]=\"true\" />\r\n                    <div class=\"input-group-append\">\r\n                      <span [style.display]=\"showClearButton ? '' : 'none'\" class=\"input-group-text input-group-clear\"\r\n                            (click)=\"clear($event)\">\r\n                        <i class=\"f-icon modal_close\"></i>\r\n                      </span>\r\n                      <span class=\"input-group-text\" (click)=\"selectUsers($event)\" style=\"background: #FFF;\">\r\n                        <i class=\"f-icon f-icon-lookup\"></i>\r\n                      </span>\r\n                    </div>\r\n                  </div>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <div class=\"d-flex flex-column\">\r\n        <div class=\"farris-form form-inline farris-form-inline mr-5\">\r\n          <div class=\"col-12 farris-group-auto my-4\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{'component.activityExecutor.enablingConditions'|wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <div class=\"input-group\">\r\n                    <input type=\"text\" class=\"form-control\" [(ngModel)]=\"condition\" name=\"conditon\" [readOnly]=\"true\" />\r\n                    <div class=\"input-group-append\">\r\n                      <span class=\"input-group-text\" (click)=\"editConditon()\" style=\"background: #FFF;\">\r\n                        <i class=\"f-icon f-icon-lookup\"></i>\r\n                      </span>\r\n                    </div>\r\n                  </div>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </form>\r\n</div>\r\n</div>\r\n\r\n<farris-dialog #allActivityDialog [buttons]=\"buttonRef\" [width]=\"700\" [height]=\"520\" title=\"{{'component.activityExecutor.activityList'|wfLocale}}\">\r\n  <div class=\"f-utils-fill-flex-column farris-overflow-y-auto h-100 m-2\">\r\n    <farris-datagrid #allActivityTable [columns]=\"cols\" [data]=\"designerUIState.userActList\"  [idField]=\"'Id'\" [fit]=\"true\"\r\n                     [fitColumns]=\"true\" [pagination]='false' [multiSelect]=\"true\" [showCheckbox]=\"true\"\r\n                     [checkOnSelect]=\"true\" [selectOnCheck]=\"true\">\r\n    </farris-datagrid>\r\n  </div>\r\n</farris-dialog>\r\n<ng-template #buttonRef>\r\n  <button class=\"btn btn-sm btn-default\" (click)=\"allActivityDialog.close()\">{{'component.activityExecutor.clickCancel'|wfLocale}}</button>\r\n  <button class=\"btn btn-sm btn-primary mx-2\" (click)=\"activitiesSelect()\">{{'component.activityExecutor.clickOK'|wfLocale}}</button>\r\n</ng-template>\r\n",
                    providers: [
                        WfLocalePipe,
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ActivityExecutorPerformerComponent.ctorParameters = function () { return [
        { type: NgZone },
        { type: ProcessDesignerUIState },
        { type: ExpressionService },
        { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
        { type: WfLocalePipe, decorators: [{ type: Optional }] }
    ]; };
    ActivityExecutorPerformerComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        inputGroup: [{ type: ViewChild, args: ['inputGroup',] }],
        allActivityDialog: [{ type: ViewChild, args: ['allActivityDialog',] }],
        allActivityTable: [{ type: ViewChild, args: ['allActivityTable',] }]
    };
    return ActivityExecutorPerformerComponent;
}());
export { ActivityExecutorPerformerComponent };
if (false) {
    /** @type {?} */
    ActivityExecutorPerformerComponent.prototype.cls;
    /** @type {?} */
    ActivityExecutorPerformerComponent.prototype.showClearButton;
    /** @type {?} */
    ActivityExecutorPerformerComponent.prototype.activityExecutors;
    /** @type {?} */
    ActivityExecutorPerformerComponent.prototype.condition;
    /** @type {?} */
    ActivityExecutorPerformerComponent.prototype.cols;
    /** @type {?} */
    ActivityExecutorPerformerComponent.prototype.inputGroup;
    /** @type {?} */
    ActivityExecutorPerformerComponent.prototype.allActivityDialog;
    /** @type {?} */
    ActivityExecutorPerformerComponent.prototype.allActivityTable;
    /**
     * @type {?}
     * @private
     */
    ActivityExecutorPerformerComponent.prototype.ngZone;
    /** @type {?} */
    ActivityExecutorPerformerComponent.prototype.designerUIState;
    /**
     * @type {?}
     * @private
     */
    ActivityExecutorPerformerComponent.prototype.expr;
    /**
     * @type {?}
     * @private
     */
    ActivityExecutorPerformerComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    ActivityExecutorPerformerComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    ActivityExecutorPerformerComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    ActivityExecutorPerformerComponent.prototype.localePipe;
}
