// @edp-bif/common-api/bundles/edp-bif-common-api.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@edp-aif/common-api')) :
    typeof define === 'function' && define.amd ? define('@edp-bif/common-api', ['exports', '@edp-aif/common-api'], factory) :
    (factory((global['edp-bif'] = global['edp-bif'] || {}, global['edp-bif']['common-api'] = {}),global.commonApi));
}(this, (function (exports,commonApi) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/ref/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/ConditionType.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var ConditionType = {
        /**
         * 增量表达式。
         */
        // INCREMENTAL_EXPRESSION = "INCREMENTAL_EXPRESSION",
        /**
         * 全量表达式。
         */
        // FULL_EXPRESSION = "FULL_EXPRESSION",
        /**
         * 表达式。
         */
        EXPRESSION: "EXPRESSION",
        /**
         * 构件。
         */
        COMPONENT: "COMPONENT",
        /**
        * VO动作。
        */
        VOACTION: "VOACTION",
        /**
         * 实现类。
         */
        IMPLCLASS: "IMPLCLASS",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/ConditionEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 条件实体类。
     */
    var /**
     * 条件实体类。
     */ ConditionEntity = /** @class */ (function (_super) {
        __extends(ConditionEntity, _super);
        function ConditionEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.isEnabled = false;
            _this.remarks = null;
            _this.supportAutoProcess = true;
            _this.supportPushProcess = false;
            _this.supportPullProcess = false;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        ConditionEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.id = null;
                this.isEnabled = false;
                this.name = null;
                this.type = null;
                this.content = null;
                this.remarks = null;
                this.supportAutoProcess = true;
                this.supportPushProcess = false;
                this.supportPullProcess = false;
            };
        /**
         * @return {?}
         */
        ConditionEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "id", this.id);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "type", this.type);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "remarks", this.remarks);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "supportAutoProcess", this.supportAutoProcess);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "supportPushProcess", this.supportPushProcess);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "supportPullProcess", this.supportPullProcess);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        ConditionEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.id = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "id");
                this.isEnabled = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
                this.name = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
                this.type = ConditionType[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "type")];
                this.content = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
                this.remarks = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "remarks");
                this.supportAutoProcess = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "supportAutoProcess");
                this.supportPushProcess = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "supportPushProcess");
                this.supportPullProcess = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "supportPullProcess");
            };
        return ConditionEntity;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/MappingMode.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var MappingMode = {
        /**
         * 映射规则。
         */
        MAPPINGRULE: "MAPPINGRULE",
        /**
         * 自定义映射构件。
         */
        COMPONENT: "COMPONENT",
        /**
         * VO动作。
         */
        VOACTION: "VOACTION",
        /**
         * 映射规则ID。
         */
        MAPPINGRULEID: "MAPPINGRULEID",
        /**
         * 表达式
         */
        EXPRESSION: "EXPRESSION",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/MappingModeEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 映射方式实体类。
     */
    var /**
     * 映射方式实体类。
     */ MappingModeEntity = /** @class */ (function (_super) {
        __extends(MappingModeEntity, _super);
        function MappingModeEntity() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        MappingModeEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.mappingMode = null;
                this.content = null;
                this.mappingRuleContent = null;
                this.componentContent = null;
                this.voActionContent = null;
                this.expressionContent = null;
            };
        /**
         * @return {?}
         */
        MappingModeEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingMode", this.mappingMode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleContent", this.mappingRuleContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "componentContent", this.componentContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "voActionContent", this.voActionContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "expressionContent", this.expressionContent);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        MappingModeEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.mappingMode = MappingMode[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingMode")];
                this.content = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
                this.mappingRuleContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleContent");
                this.componentContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentContent");
                this.voActionContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voActionContent");
                this.expressionContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "expressionContent");
            };
        return MappingModeEntity;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/SaveMode.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var SaveMode = {
        /**
         * 默认。
         */
        DEFAULT: "DEFAULT",
        /**
         * 构件。
         */
        COMPONENT: "COMPONENT",
        /**
         * Vo动作。
         */
        VOACTION: "VOACTION",
        /**
         * 任务中心
         */
        TASKCENTER: "TASKCENTER",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/SaveModeEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 保存模式实体类。
     */
    var /**
     * 保存模式实体类。
     */ SaveModeEntity = /** @class */ (function (_super) {
        __extends(SaveModeEntity, _super);
        function SaveModeEntity() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        SaveModeEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.saveMode = null;
                this.content = null;
                this.componentContent = null;
                this.voActionContent = null;
                this.taskCenterContent = null;
            };
        /**
         * @return {?}
         */
        SaveModeEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "saveMode", this.saveMode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "componentContent", this.componentContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "voActionContent", this.voActionContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterContent", this.taskCenterContent);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        SaveModeEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.saveMode = SaveMode[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveMode")];
                this.content = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
                this.componentContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentContent");
                this.voActionContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voActionContent");
                this.taskCenterContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterContent");
            };
        return SaveModeEntity;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/content/impl/AbstractRtBillProcessContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 单据流内容类。
     * @param TEntity 单据流实体。
     * @abstract
     * @template TEntity
     */
    var /**
     * 单据流内容类。
     * @param TEntity 单据流实体。
     * @abstract
     * @template TEntity
     */ AbstractRtBillProcessContent = /** @class */ (function (_super) {
        __extends(AbstractRtBillProcessContent, _super);
        function AbstractRtBillProcessContent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        AbstractRtBillProcessContent.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.entity = null;
                this.eventRouterConditionsContent = new Array();
                this.mappingConditionsContent = new Array();
                this.mappingModeContent = null;
                this.saveModeContent = null;
            };
        /**
         * @return {?}
         */
        AbstractRtBillProcessContent.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "entity", this.entity);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsContent", this.eventRouterConditionsContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsContent", this.mappingConditionsContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeContent", this.mappingModeContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeContent", this.saveModeContent);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        AbstractRtBillProcessContent.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.entity = commonApi.BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "entity", this.getEntityClass());
                this.eventRouterConditionsContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(obj, "eventRouterConditionsContent", ConditionEntity);
                this.mappingConditionsContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingConditionsContent", ConditionEntity);
                this.mappingModeContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "mappingModeContent", MappingModeEntity);
                this.saveModeContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "saveModeContent", SaveModeEntity);
            };
        return AbstractRtBillProcessContent;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/AbstractBillProcessEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 单据流实体类。
     */
    var /**
     * 单据流实体类。
     */ AbstractBillProcessEntity = /** @class */ (function (_super) {
        __extends(AbstractBillProcessEntity, _super);
        function AbstractBillProcessEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.enabledParentEventRouterCond = true;
            _this.enabledParentMappingCond = true;
            _this.sysInit = false;
            _this.origin = true;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        AbstractBillProcessEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.code = null;
                this.name = null;
                this.description = null;
                this.persistenceVersion = null;
                this.version = 0;
                this.callType = null;
                // this.bizTypeL1 = null;
                // this.bizTypeL2 = null;
                // this.bizTypeL3 = null;
                // this.bizTypeL4 = null;
                this.upBillAppCode = null;
                this.upBillSuCode = null;
                this.upBillBoId = null;
                this.upBillBeMetadataId = null;
                this.upBillBeId = null;
                this.upBillVoMetadataId = null;
                this.upBillVoId = null;
                this.downBillAppCode = null;
                this.downBillSuCode = null;
                this.downBillBoId = null;
                this.downBillBeMetadataId = null;
                this.downBillBeId = null;
                this.downBillVoMetadataId = null;
                this.downBillVoId = null;
                this.isSupportAutomatinFlow = false;
                this.bizEventId = null;
                this.bizActionId = null;
                this.originId = null;
                this.rootId = null;
                this.parentId = null;
                this.enabledParentEventRouterCond = true;
                this.enabledParentMappingCond = true;
                this.idPath = null;
                this.versionTag = null;
                this.upBillBeMetadataRootId = null;
                this.upBillBeMetadataParentId = null;
                this.upBillBeMetadataIdPath = null;
                this.upBillBeRootId = null;
                this.upBillBeParentId = null;
                this.upBillBeIdPath = null;
                this.upBillVoMetadataRootId = null;
                this.upBillVoMetadataParentId = null;
                this.upBillVoMetadataIdPath = null;
                this.upBillVoRootId = null;
                this.upBillVoParentId = null;
                this.upBillVoIdPath = null;
                this.downBillBeMetadataRootId = null;
                this.downBillBeMetadataParentId = null;
                this.downBillBeMetadataIdPath = null;
                this.downBillBeRootId = null;
                this.downBillBeParentId = null;
                this.downBillBeIdPath = null;
                this.downBillVoMetadataRootId = null;
                this.downBillVoMetadataParentId = null;
                this.downBillVoMetadataIdPath = null;
                this.downBillVoRootId = null;
                this.downBillVoParentId = null;
                this.downBillVoIdPath = null;
                this.sysInit = false;
                this.origin = true;
                this.colI18nValues = null;
            };
        /**
         * @return {?}
         */
        AbstractBillProcessEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "description", this.description);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "persistenceVersion", this.persistenceVersion);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "version", this.version);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "callType", this.callType);
                // BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeL1", this.bizTypeL1);
                // BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeL2", this.bizTypeL2);
                // BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeL3", this.bizTypeL3);
                // BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeL4", this.bizTypeL4);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillAppCode", this.upBillAppCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillSuCode", this.upBillSuCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBoId", this.upBillBoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeMetadataId", this.upBillBeMetadataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeId", this.upBillBeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataId", this.upBillVoMetadataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoId", this.upBillVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillAppCode", this.downBillAppCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillSuCode", this.downBillSuCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBoId", this.downBillBoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeMetadataId", this.downBillBeMetadataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeId", this.downBillBeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataId", this.downBillVoMetadataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoId", this.downBillVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportAutomatinFlow", this.isSupportAutomatinFlow);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventId", this.bizEventId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionId", this.bizActionId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "originId", this.originId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "rootId", this.rootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "parentId", this.parentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "enabledParentEventRouterCond", this.enabledParentEventRouterCond);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "enabledParentMappingCond", this.enabledParentMappingCond);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "idPath", this.idPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "versionTag", this.versionTag);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeMetadataRootId", this.upBillBeMetadataRootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeMetadataParentId", this.upBillBeMetadataParentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeMetadataIdPath", this.upBillBeMetadataIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeRootId", this.upBillBeRootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeParentId", this.upBillBeParentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeIdPath", this.upBillBeIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataRootId", this.upBillVoMetadataRootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataParentId", this.upBillVoMetadataParentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataIdPath", this.upBillVoMetadataIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoRootId", this.upBillVoRootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoParentId", this.upBillVoParentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoIdPath", this.upBillVoIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeMetadataRootId", this.downBillBeMetadataRootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeMetadataParentId", this.downBillBeMetadataParentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeMetadataIdPath", this.downBillBeMetadataIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeRootId", this.downBillBeRootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeParentId", this.downBillBeParentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeIdPath", this.downBillBeIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataRootId", this.downBillVoMetadataRootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataParentId", this.downBillVoMetadataParentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataIdPath", this.downBillVoMetadataIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoRootId", this.downBillVoRootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoParentId", this.downBillVoParentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoIdPath", this.downBillVoIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sysInit", this.sysInit);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "origin", this.origin);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "colI18nValues", this.colI18nValues);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        AbstractBillProcessEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.code = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
                this.name = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
                this.description = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "description");
                this.persistenceVersion = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "persistenceVersion");
                this.version = commonApi.BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "version");
                this.callType = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "callType");
                // this.bizTypeL1 = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeL1");
                // this.bizTypeL2 = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeL2");
                // this.bizTypeL3 = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeL3");
                // this.bizTypeL4 = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeL4");
                this.upBillAppCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillAppCode");
                this.upBillSuCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillSuCode");
                this.upBillBoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBoId");
                this.upBillBeMetadataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeMetadataId");
                this.upBillBeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeId");
                this.upBillVoMetadataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataId");
                this.upBillVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoId");
                this.downBillAppCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillAppCode");
                this.downBillSuCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillSuCode");
                this.downBillBoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBoId");
                this.downBillBeMetadataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeMetadataId");
                this.downBillBeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeId");
                this.downBillVoMetadataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataId");
                this.downBillVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoId");
                this.isSupportAutomatinFlow = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportAutomatinFlow");
                this.bizEventId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventId");
                this.bizActionId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionId");
                this.originId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originId");
                this.rootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "rootId");
                this.parentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "parentId");
                this.enabledParentEventRouterCond = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "enabledParentEventRouterCond");
                this.enabledParentMappingCond = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "enabledParentMappingCond");
                this.idPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "idPath");
                this.versionTag = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "versionTag");
                this.upBillBeMetadataRootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeMetadataRootId");
                this.upBillBeMetadataParentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeMetadataParentId");
                this.upBillBeMetadataIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeMetadataIdPath");
                this.upBillBeRootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeRootId");
                this.upBillBeParentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeParentId");
                this.upBillBeIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeIdPath");
                this.upBillVoMetadataRootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataRootId");
                this.upBillVoMetadataParentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataParentId");
                this.upBillVoMetadataIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataIdPath");
                this.upBillVoRootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoRootId");
                this.upBillVoParentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoParentId");
                this.upBillVoIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoIdPath");
                this.downBillBeMetadataRootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeMetadataRootId");
                this.downBillBeMetadataParentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeMetadataParentId");
                this.downBillBeMetadataIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeMetadataIdPath");
                this.downBillBeRootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeRootId");
                this.downBillBeParentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeParentId");
                this.downBillBeIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeIdPath");
                this.downBillVoMetadataRootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataRootId");
                this.downBillVoMetadataParentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataParentId");
                this.downBillVoMetadataIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataIdPath");
                this.downBillVoRootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoRootId");
                this.downBillVoParentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoParentId");
                this.downBillVoIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoIdPath");
                this.sysInit = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "sysInit");
                this.origin = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "origin");
                this.colI18nValues = commonApi.BpmnModelJsonHelper.GetAttributeValue(obj, "colI18nValues");
            };
        return AbstractBillProcessEntity;
    }(commonApi.Entity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/AbstractBillCreationProcessEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 单据生单流实体类。
     */
    var /**
     * 单据生单流实体类。
     */ AbstractBillCreationProcessEntity = /** @class */ (function (_super) {
        __extends(AbstractBillCreationProcessEntity, _super);
        function AbstractBillCreationProcessEntity() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        AbstractBillCreationProcessEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.creationRuleId = null;
                this.creationProcessId = null;
                this.downBillVoProxyMetadataId = null;
                this.downBillVoProxyId = null;
                this.downBillVoProxyMetadataRootId = null;
                this.downBillVoProxyMetadataParId = null;
                this.downBillVoProxyMetadataIdPath = null;
                this.downBillVoProxyRootId = null;
                this.downBillVoProxyParentId = null;
                this.downBillVoProxyIdPath = null;
            };
        /**
         * @return {?}
         */
        AbstractBillCreationProcessEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessId", this.creationProcessId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataId", this.downBillVoProxyMetadataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyId", this.downBillVoProxyId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataRootId", this.downBillVoProxyMetadataRootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataParId", this.downBillVoProxyMetadataParId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataIdPath", this.downBillVoProxyMetadataIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyRootId", this.downBillVoProxyRootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyParentId", this.downBillVoProxyParentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyIdPath", this.downBillVoProxyIdPath);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        AbstractBillCreationProcessEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.creationRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
                this.creationProcessId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessId");
                this.downBillVoProxyMetadataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataId");
                this.downBillVoProxyId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyId");
                this.downBillVoProxyMetadataRootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataRootId");
                this.downBillVoProxyMetadataParId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataParId");
                this.downBillVoProxyMetadataIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataIdPath");
                this.downBillVoProxyRootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyRootId");
                this.downBillVoProxyParentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyParentId");
                this.downBillVoProxyIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyIdPath");
            };
        return AbstractBillCreationProcessEntity;
    }(AbstractBillProcessEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/RtBillCreationProcessEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RtBillCreationProcessEntity = /** @class */ (function (_super) {
        __extends(RtBillCreationProcessEntity, _super);
        function RtBillCreationProcessEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.isEnabled = false;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        RtBillCreationProcessEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.isEnabled = false;
                this.eventRouterConditionsId = null;
                this.mappingConditionsId = null;
                this.mappingModeId = null;
                this.saveModeId = null;
                this.formSettingsId = null;
                this.dtBillCreationProcessId = null;
                this.upBillTrackingSettingsId = null;
                this.downBillTrackingSettingsId = null;
                this.reverseWriteRuleId = null;
                this.taskCenterConfig4Push = null;
                this.taskCenterConfig4Pull = null;
                this.isSupportPull = false;
                this.isSupportPush = false;
            };
        /**
         * @return {?}
         */
        RtBillCreationProcessEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsId", this.eventRouterConditionsId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsId", this.mappingConditionsId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeId", this.mappingModeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeId", this.saveModeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "formSettingsId", this.formSettingsId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "dtBillCreationProcessId", this.dtBillCreationProcessId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettingsId", this.upBillTrackingSettingsId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettingsId", this.downBillTrackingSettingsId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleId", this.reverseWriteRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Push", this.taskCenterConfig4Push);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Pull", this.taskCenterConfig4Pull);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPull", this.isSupportPull);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPush", this.isSupportPush);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RtBillCreationProcessEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.isEnabled = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
                this.eventRouterConditionsId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsId");
                this.mappingConditionsId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsId");
                this.mappingModeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeId");
                this.saveModeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeId");
                this.formSettingsId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettingsId");
                this.dtBillCreationProcessId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dtBillCreationProcessId");
                this.upBillTrackingSettingsId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettingsId");
                this.downBillTrackingSettingsId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettingsId");
                this.reverseWriteRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleId");
                this.taskCenterConfig4Push = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Push");
                this.taskCenterConfig4Pull = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Pull");
                this.isSupportPull = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPull");
                this.isSupportPush = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPush");
            };
        return RtBillCreationProcessEntity;
    }(AbstractBillCreationProcessEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/content/impl/RtBillCreationProcessContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 单据生单流内容类。
     */
    var /**
     * 单据生单流内容类。
     */ RtBillCreationProcessContent = /** @class */ (function (_super) {
        __extends(RtBillCreationProcessContent, _super);
        function RtBillCreationProcessContent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        RtBillCreationProcessContent.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.formSettings = null;
                this.upBillTrackingSettings = null;
                this.downBillTrackingSettings = null;
                this.reverseWriteRuleContent = null;
                this.taskCenterConfig4Push = null;
                this.taskCenterConfig4Pull = null;
            };
        /**
         * @return {?}
         */
        RtBillCreationProcessContent.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "formSettings", this.formSettings);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettings", this.upBillTrackingSettings);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettings", this.downBillTrackingSettings);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleContent", this.reverseWriteRuleContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Push", this.taskCenterConfig4Push);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Pull", this.taskCenterConfig4Pull);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RtBillCreationProcessContent.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.formSettings = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettings");
                this.upBillTrackingSettings = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettings");
                this.downBillTrackingSettings = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettings");
                this.reverseWriteRuleContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleContent");
                this.taskCenterConfig4Push = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Push");
                this.taskCenterConfig4Pull = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Pull");
            };
        /**
         * @protected
         * @return {?}
         */
        RtBillCreationProcessContent.prototype.getEntityClass = /**
         * @protected
         * @return {?}
         */
            function () {
                return RtBillCreationProcessEntity;
            };
        return RtBillCreationProcessContent;
    }(AbstractRtBillProcessContent));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/AbstractBillWritebackProcessEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 单据回写流实体类。
     */
    var /**
     * 单据回写流实体类。
     */ AbstractBillWritebackProcessEntity = /** @class */ (function (_super) {
        __extends(AbstractBillWritebackProcessEntity, _super);
        function AbstractBillWritebackProcessEntity() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        AbstractBillWritebackProcessEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.writebackRuleId = null;
                this.writebackProcessId = null;
                this.upBillVoProxyMetadataId = null;
                this.upBillVoProxyId = null;
                this.upBillVoProxyMetadataRootId = null;
                this.upBillVoProxyMetadataParentId = null;
                this.upBillVoProxyMetadataIdPath = null;
                this.upBillVoProxyRootId = null;
                this.upBillVoProxyParentId = null;
                this.upBillVoProxyIdPath = null;
            };
        /**
         * @return {?}
         */
        AbstractBillWritebackProcessEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "writebackRuleId", this.writebackRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "writebackProcessId", this.writebackProcessId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyMetadataId", this.upBillVoProxyMetadataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyId", this.upBillVoProxyId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyMetadataRootId", this.upBillVoProxyMetadataRootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyMetadataParentId", this.upBillVoProxyMetadataParentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyMetadataIdPath", this.upBillVoProxyMetadataIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyRootId", this.upBillVoProxyRootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyParentId", this.upBillVoProxyParentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyIdPath", this.upBillVoProxyIdPath);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        AbstractBillWritebackProcessEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.writebackRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackRuleId");
                this.writebackProcessId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackProcessId");
                this.upBillVoProxyMetadataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyMetadataId");
                this.upBillVoProxyId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyId");
                this.upBillVoProxyMetadataRootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyMetadataRootId");
                this.upBillVoProxyMetadataParentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyMetadataParentId");
                this.upBillVoProxyMetadataIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyMetadataIdPath");
                this.upBillVoProxyRootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyRootId");
                this.upBillVoProxyParentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyParentId");
                this.upBillVoProxyIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyIdPath");
            };
        return AbstractBillWritebackProcessEntity;
    }(AbstractBillProcessEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/RtBillWritebackProcessEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RtBillWritebackProcessEntity = /** @class */ (function (_super) {
        __extends(RtBillWritebackProcessEntity, _super);
        function RtBillWritebackProcessEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.isEnabled = false;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        RtBillWritebackProcessEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.isEnabled = false;
                this.eventRouterConditionsId = null;
                this.mappingConditionsId = null;
                this.mappingModeId = null;
                this.saveModeId = null;
                this.dtBillWritebackProcessId = null;
            };
        /**
         * @return {?}
         */
        RtBillWritebackProcessEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsId", this.eventRouterConditionsId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsId", this.mappingConditionsId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeId", this.mappingModeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeId", this.saveModeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "dtBillWritebackProcessId", this.dtBillWritebackProcessId);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RtBillWritebackProcessEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.isEnabled = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
                this.eventRouterConditionsId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsId");
                this.mappingConditionsId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsId");
                this.mappingModeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeId");
                this.saveModeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeId");
                this.dtBillWritebackProcessId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dtBillWritebackProcessId");
            };
        return RtBillWritebackProcessEntity;
    }(AbstractBillWritebackProcessEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/content/impl/RtBillWritebackProcessContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 单据回写流内容类。
     */
    var /**
     * 单据回写流内容类。
     */ RtBillWritebackProcessContent = /** @class */ (function (_super) {
        __extends(RtBillWritebackProcessContent, _super);
        function RtBillWritebackProcessContent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        RtBillWritebackProcessContent.prototype.getEntityClass = /**
         * @protected
         * @return {?}
         */
            function () {
                return RtBillWritebackProcessEntity;
            };
        return RtBillWritebackProcessContent;
    }(AbstractRtBillProcessContent));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/content/impl/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/content/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/BillTrackingDataShownMode.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var BillTrackingDataShownMode = {
        /**
        * 默认。
        */
        DEFAULT: "DEFAULT",
        /**
         * 自定义显示字段。
         */
        CUSTOMFIELD: "CUSTOMFIELD",
        /**
         * 自定义表单。
         */
        CUSTOMFORM: "CUSTOMFORM",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/BillTrackingNodeShownMode.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var BillTrackingNodeShownMode = {
        /**
        * 默认。
        */
        DEFAULT: "DEFAULT",
        /**
         * 自定义。
         */
        CUSTOM: "CUSTOM",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/JointQueryMode.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var JointQueryMode = {
        /**
         * 默认。
         */
        DEFAULT: "DEFAULT",
        /**
         * 通过App联查。
         */
        APP: "APP",
        /**
         * 通过URL联查。
         */
        URL: "URL",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/TaskCenterConfig.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var TaskCenterConfig = /** @class */ (function (_super) {
        __extends(TaskCenterConfig, _super);
        function TaskCenterConfig() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.jointQueryMode = JointQueryMode.APP;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        TaskCenterConfig.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.taskCenterContent = null;
                this.jointQueryMode = JointQueryMode.APP;
                this.joitQueryMenu = null;
                this.jointQueryUrl = null;
            };
        /**
         * @return {?}
         */
        TaskCenterConfig.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterContent", this.taskCenterContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryMode", this.jointQueryMode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "joitQueryMenu", this.joitQueryMenu);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryUrl", this.jointQueryUrl);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        TaskCenterConfig.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.taskCenterContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "taskCenterContent", commonApi.TaskCenterContent);
                this.jointQueryMode = JointQueryMode[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryMode")];
                this.joitQueryMenu = commonApi.BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "joitQueryMenu", commonApi.TaskHandlingMenuArgs);
                this.jointQueryUrl = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryUrl");
            };
        return TaskCenterConfig;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/TaskCenterConfig4Pull.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var TaskCenterConfig4Pull = /** @class */ (function (_super) {
        __extends(TaskCenterConfig4Pull, _super);
        function TaskCenterConfig4Pull() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return TaskCenterConfig4Pull;
    }(TaskCenterConfig));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/TaskCenterConfig4Push.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var TaskCenterConfig4Push = /** @class */ (function (_super) {
        __extends(TaskCenterConfig4Push, _super);
        function TaskCenterConfig4Push() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return TaskCenterConfig4Push;
    }(TaskCenterConfig));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/AbstractBillCpAndWpEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var AbstractBillCpAndWpEntity = /** @class */ (function (_super) {
        __extends(AbstractBillCpAndWpEntity, _super);
        function AbstractBillCpAndWpEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.sysInit = false;
            _this.isCreationProcess = false;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        AbstractBillCpAndWpEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.code = null;
                this.name = null;
                this.persistenceVersion = null;
                this.parentId = null;
                this.upBillVoId = null;
                this.downBillVoId = null;
                this.sysInit = false;
                this.isCreationProcess = false;
                this.originId = null;
                this.rootId = null;
                this.colI18nValues = null;
            };
        /**
         * @return {?}
         */
        AbstractBillCpAndWpEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "persistenceVersion", this.persistenceVersion);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "parentId", this.parentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoId", this.upBillVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoId", this.downBillVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sysInit", this.sysInit);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isCreationProcess", this.isCreationProcess);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "originId", this.originId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "rootId", this.rootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "colI18nValues", this.colI18nValues);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        AbstractBillCpAndWpEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.code = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
                this.name = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
                this.persistenceVersion = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "persistenceVersion");
                this.parentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "parentId");
                this.upBillVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoId");
                this.downBillVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoId");
                this.sysInit = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "sysInit");
                this.isCreationProcess = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isCreationProcess");
                this.originId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originId");
                this.rootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "rootId");
                this.colI18nValues = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "colI18nValues");
            };
        return AbstractBillCpAndWpEntity;
    }(commonApi.Entity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/BillTrackingConfigEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BillTrackingConfigEntity = /** @class */ (function (_super) {
        __extends(BillTrackingConfigEntity, _super);
        function BillTrackingConfigEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.isEnabled = false;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        BillTrackingConfigEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.suCode = null;
                this.classTypeName = null;
                this.isEnabled = false;
            };
        /**
         * @return {?}
         */
        BillTrackingConfigEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "classTypeName", this.classTypeName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        BillTrackingConfigEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.suCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
                this.classTypeName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "classTypeName");
                this.isEnabled = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
            };
        return BillTrackingConfigEntity;
    }(commonApi.Entity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/BillTrackingSettings.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BillTrackingSettings = /** @class */ (function (_super) {
        __extends(BillTrackingSettings, _super);
        function BillTrackingSettings() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.billTrackingNodeShownMode = BillTrackingNodeShownMode.DEFAULT;
            _this.billTrackingDataShownMode = BillTrackingDataShownMode.DEFAULT;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        BillTrackingSettings.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.billTrackingNodeShownMode = BillTrackingNodeShownMode.DEFAULT;
                this.billNodeName = null;
                this.billNodeShownStyle = null;
                this.summaryFieldsInfo = null;
                this.billTrackingDataShownMode = BillTrackingDataShownMode.DEFAULT;
                this.customFieldContent = null;
                this.customFormContent = null;
                this.version = null;
            };
        /**
         * @return {?}
         */
        BillTrackingSettings.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "billNodeName", this.billNodeName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "billTrackingNodeShownMode", this.billTrackingNodeShownMode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "billNodeShownStyle", this.billNodeShownStyle);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "summaryFieldsInfo", this.summaryFieldsInfo);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "billTrackingDataShownMode", this.billTrackingDataShownMode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "customFieldContent", this.customFieldContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "customFormContent", this.customFormContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "version", this.version);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        BillTrackingSettings.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.billNodeName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billNodeName");
                this.billTrackingNodeShownMode = BillTrackingDataShownMode[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billTrackingNodeShownMode")];
                this.billNodeShownStyle = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billNodeShownStyle");
                this.summaryFieldsInfo = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "summaryFieldsInfo");
                this.billTrackingDataShownMode = BillTrackingDataShownMode[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billTrackingDataShownMode")];
                this.customFieldContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "customFieldContent");
                this.customFormContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "customFormContent");
                this.version = ( /** @type {?} */(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "version")));
            };
        return BillTrackingSettings;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/ConditionContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ConditionContent = /** @class */ (function (_super) {
        __extends(ConditionContent, _super);
        function ConditionContent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        ConditionContent.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.unSatisfiedInformation = null;
            };
        /**
         * @return {?}
         */
        ConditionContent.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "unSatisfiedInformation", this.unSatisfiedInformation);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        ConditionContent.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.unSatisfiedInformation = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "unSatisfiedInformation");
            };
        return ConditionContent;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/ComponentConditionContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ComponentConditionContent = /** @class */ (function (_super) {
        __extends(ComponentConditionContent, _super);
        function ComponentConditionContent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        ComponentConditionContent.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.componentMetadataId = null;
                this.componentId = null;
                this.name = null;
            };
        /**
         * @return {?}
         */
        ComponentConditionContent.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "componentMetadataId", this.componentMetadataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "componentId", this.componentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        ComponentConditionContent.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.componentMetadataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentMetadataId");
                this.componentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentId");
                this.name = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
            };
        return ComponentConditionContent;
    }(ConditionContent));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/ComponentContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ComponentContent = /** @class */ (function (_super) {
        __extends(ComponentContent, _super);
        function ComponentContent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        ComponentContent.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.componentMetadataId = null;
                this.componentId = null;
                this.name = null;
            };
        /**
         * @return {?}
         */
        ComponentContent.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "componentMetadataId", this.componentMetadataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "componentId", this.componentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        ComponentContent.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.componentMetadataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentMetadataId");
                this.componentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentId");
                this.name = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
            };
        return ComponentContent;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/DtBillCpAndWpEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var DtBillCpAndWpEntity = /** @class */ (function (_super) {
        __extends(DtBillCpAndWpEntity, _super);
        function DtBillCpAndWpEntity() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return DtBillCpAndWpEntity;
    }(AbstractBillCpAndWpEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/DtBillCreationProcessEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var DtBillCreationProcessEntity = /** @class */ (function (_super) {
        __extends(DtBillCreationProcessEntity, _super);
        function DtBillCreationProcessEntity() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        DtBillCreationProcessEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.eventRouterConditionsContent = null;
                this.mappingConditionsContent = null;
                this.mappingModeContent = null;
                this.saveModeContent = null;
                this.formSettingsContent = null;
                this.upBillTrackingSettings = null;
                this.downBillTrackingSettings = null;
                this.reverseWriteRuleContent = null;
                this.taskCenterConfig4Push = null;
                this.taskCenterConfig4Pull = null;
                this.isSupportPull = false;
                this.isSupportPush = false;
            };
        /**
         * @return {?}
         */
        DtBillCreationProcessEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsContent", this.eventRouterConditionsContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsContent", this.mappingConditionsContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeContent", this.mappingModeContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeContent", this.saveModeContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "formSettingsContent", this.formSettingsContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettings", this.upBillTrackingSettings);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettings", this.downBillTrackingSettings);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleContent", this.reverseWriteRuleContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Push", this.taskCenterConfig4Push);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Pull", this.taskCenterConfig4Pull);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPull", this.isSupportPull);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPush", this.isSupportPush);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        DtBillCreationProcessEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.eventRouterConditionsContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsContent");
                this.mappingConditionsContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsContent");
                this.mappingModeContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeContent");
                this.saveModeContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeContent");
                this.formSettingsContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettingsContent");
                this.upBillTrackingSettings = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettings");
                this.downBillTrackingSettings = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettings");
                this.reverseWriteRuleContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleContent");
                this.taskCenterConfig4Push = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Push");
                this.taskCenterConfig4Pull = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Pull");
                this.isSupportPull = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPull");
                this.isSupportPush = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPush");
            };
        return DtBillCreationProcessEntity;
    }(AbstractBillCreationProcessEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/DtBillWritebackProcessEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var DtBillWritebackProcessEntity = /** @class */ (function (_super) {
        __extends(DtBillWritebackProcessEntity, _super);
        function DtBillWritebackProcessEntity() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        DtBillWritebackProcessEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.eventRouterConditionsContent = null;
                this.mappingConditionsContent = null;
                this.mappingModeContent = null;
                this.saveModeContent = null;
            };
        /**
         * @return {?}
         */
        DtBillWritebackProcessEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsContent", this.eventRouterConditionsContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsContent", this.mappingConditionsContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeContent", this.mappingModeContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeContent", this.saveModeContent);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        DtBillWritebackProcessEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.eventRouterConditionsContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsContent");
                this.mappingConditionsContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsContent");
                this.mappingModeContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeContent");
                this.saveModeContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeContent");
            };
        return DtBillWritebackProcessEntity;
    }(AbstractBillWritebackProcessEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/ExportLogResult.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ExportLogResult = /** @class */ (function (_super) {
        __extends(ExportLogResult, _super);
        function ExportLogResult() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.aifLogEntities = new Array();
            _this.rtBillCreationProcessEntities = new Array();
            _this.rtBillWritebackProcessEntities = new Array();
            _this.bizEventEntities = new Array();
            _this.bizActionEntities = new Array();
            _this.creationRuleEntities = new Array();
            _this.creationProcessEntities = new Array();
            _this.writebackRuleEntities = new Array();
            _this.writebackProcessEntities = new Array();
            _this.generalTextEntities = new Array();
            _this.aifLogEntitiesMap = {};
            _this.rtBillCreationProcessEntitiesMap = {};
            _this.rtBillWritebackProcessEntitiesMap = {};
            _this.bizEventEntitiesMap = {};
            _this.bizActionEntitiesMap = {};
            _this.creationRuleEntitiesMap = {};
            _this.creationProcessEntitiesMap = {};
            _this.writebackRuleEntitiesMap = {};
            _this.writebackProcessEntitiesMap = {};
            _this.generalTextEntitiesMap = {};
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        ExportLogResult.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.aifLogEntities = [];
                this.rtBillCreationProcessEntities = [];
                this.rtBillWritebackProcessEntities = [];
                this.bizEventEntities = [];
                this.bizActionEntities = [];
                this.creationRuleEntities = [];
                this.creationProcessEntities = [];
                this.writebackRuleEntities = [];
                this.writebackProcessEntities = [];
                this.generalTextEntities = [];
                this.aifLogEntitiesMap = {};
                this.rtBillCreationProcessEntitiesMap = {};
                this.rtBillWritebackProcessEntitiesMap = {};
                this.bizEventEntitiesMap = {};
                this.bizActionEntitiesMap = {};
                this.creationRuleEntitiesMap = {};
                this.creationProcessEntitiesMap = {};
                this.writebackRuleEntitiesMap = {};
                this.writebackProcessEntitiesMap = {};
                this.generalTextEntitiesMap = {};
            };
        /**
         * @return {?}
         */
        ExportLogResult.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "aifLogEntities", commonApi.BpmnModelJsonHelper.ToJson(this.aifLogEntities));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "rtBillCreationProcessEntities", commonApi.BpmnModelJsonHelper.ToJson(this.rtBillCreationProcessEntities));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "rtBillWritebackProcessEntities", commonApi.BpmnModelJsonHelper.ToJson(this.rtBillWritebackProcessEntities));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventEntities", commonApi.BpmnModelJsonHelper.ToJson(this.bizEventEntities));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionEntities", commonApi.BpmnModelJsonHelper.ToJson(this.bizActionEntities));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleEntities", commonApi.BpmnModelJsonHelper.ToJson(this.creationRuleEntities));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessEntities", commonApi.BpmnModelJsonHelper.ToJson(this.creationProcessEntities));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "writebackRuleEntities", commonApi.BpmnModelJsonHelper.ToJson(this.writebackRuleEntities));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "writebackProcessEntities", commonApi.BpmnModelJsonHelper.ToJson(this.writebackProcessEntities));
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "generalTextEntities", commonApi.BpmnModelJsonHelper.ToJson(this.generalTextEntities));
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        ExportLogResult.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                var e_1, _a, e_2, _b, e_3, _c, e_4, _d, e_5, _e, e_6, _f, e_7, _g, e_8, _h, e_9, _j, e_10, _k;
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.aifLogEntities = commonApi.BpmnModelJsonHelper.FromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "aifLogEntities"), commonApi.AifLogEntity);
                this.rtBillCreationProcessEntities = commonApi.BpmnModelJsonHelper.FromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "rtBillCreationProcessEntities"), RtBillCreationProcessEntity);
                this.rtBillWritebackProcessEntities = commonApi.BpmnModelJsonHelper.FromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "rtBillWritebackProcessEntities"), RtBillWritebackProcessEntity);
                this.bizEventEntities = commonApi.BpmnModelJsonHelper.FromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventEntities"), commonApi.BizEventEntity);
                this.bizActionEntities = commonApi.BpmnModelJsonHelper.FromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionEntities"), commonApi.BizActionEntity);
                this.creationRuleEntities = commonApi.BpmnModelJsonHelper.FromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleEntities"), commonApi.CreationRuleEntity);
                this.creationProcessEntities = commonApi.BpmnModelJsonHelper.FromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessEntities"), commonApi.CreationProcessEntity);
                this.writebackRuleEntities = commonApi.BpmnModelJsonHelper.FromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackRuleEntities"), commonApi.WritebackRuleEntity);
                this.writebackProcessEntities = commonApi.BpmnModelJsonHelper.FromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackProcessEntities"), commonApi.WritebackProcessEntity);
                this.generalTextEntities = commonApi.BpmnModelJsonHelper.FromJson(commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "generalTextEntities"), commonApi.GeneralTextEntity);
                try {
                    for (var _l = __values(this.aifLogEntities), _m = _l.next(); !_m.done; _m = _l.next()) {
                        var entity = _m.value;
                        this.aifLogEntitiesMap[entity.id] = entity;
                    }
                }
                catch (e_1_1) {
                    e_1 = { error: e_1_1 };
                }
                finally {
                    try {
                        if (_m && !_m.done && (_a = _l.return))
                            _a.call(_l);
                    }
                    finally {
                        if (e_1)
                            throw e_1.error;
                    }
                }
                try {
                    for (var _o = __values(this.rtBillCreationProcessEntities), _p = _o.next(); !_p.done; _p = _o.next()) {
                        var entity = _p.value;
                        this.rtBillCreationProcessEntitiesMap[entity.id] = entity;
                    }
                }
                catch (e_2_1) {
                    e_2 = { error: e_2_1 };
                }
                finally {
                    try {
                        if (_p && !_p.done && (_b = _o.return))
                            _b.call(_o);
                    }
                    finally {
                        if (e_2)
                            throw e_2.error;
                    }
                }
                try {
                    for (var _q = __values(this.rtBillWritebackProcessEntities), _r = _q.next(); !_r.done; _r = _q.next()) {
                        var entity = _r.value;
                        this.rtBillWritebackProcessEntitiesMap[entity.id] = entity;
                    }
                }
                catch (e_3_1) {
                    e_3 = { error: e_3_1 };
                }
                finally {
                    try {
                        if (_r && !_r.done && (_c = _q.return))
                            _c.call(_q);
                    }
                    finally {
                        if (e_3)
                            throw e_3.error;
                    }
                }
                try {
                    for (var _s = __values(this.bizEventEntities), _t = _s.next(); !_t.done; _t = _s.next()) {
                        var entity = _t.value;
                        this.bizEventEntitiesMap[entity.id] = entity;
                    }
                }
                catch (e_4_1) {
                    e_4 = { error: e_4_1 };
                }
                finally {
                    try {
                        if (_t && !_t.done && (_d = _s.return))
                            _d.call(_s);
                    }
                    finally {
                        if (e_4)
                            throw e_4.error;
                    }
                }
                try {
                    for (var _u = __values(this.bizActionEntities), _v = _u.next(); !_v.done; _v = _u.next()) {
                        var entity = _v.value;
                        this.bizActionEntitiesMap[entity.id] = entity;
                    }
                }
                catch (e_5_1) {
                    e_5 = { error: e_5_1 };
                }
                finally {
                    try {
                        if (_v && !_v.done && (_e = _u.return))
                            _e.call(_u);
                    }
                    finally {
                        if (e_5)
                            throw e_5.error;
                    }
                }
                try {
                    for (var _w = __values(this.creationRuleEntities), _x = _w.next(); !_x.done; _x = _w.next()) {
                        var entity = _x.value;
                        this.creationRuleEntitiesMap[entity.id] = entity;
                    }
                }
                catch (e_6_1) {
                    e_6 = { error: e_6_1 };
                }
                finally {
                    try {
                        if (_x && !_x.done && (_f = _w.return))
                            _f.call(_w);
                    }
                    finally {
                        if (e_6)
                            throw e_6.error;
                    }
                }
                try {
                    for (var _y = __values(this.creationProcessEntities), _z = _y.next(); !_z.done; _z = _y.next()) {
                        var entity = _z.value;
                        this.creationProcessEntitiesMap[entity.id] = entity;
                    }
                }
                catch (e_7_1) {
                    e_7 = { error: e_7_1 };
                }
                finally {
                    try {
                        if (_z && !_z.done && (_g = _y.return))
                            _g.call(_y);
                    }
                    finally {
                        if (e_7)
                            throw e_7.error;
                    }
                }
                try {
                    for (var _0 = __values(this.writebackRuleEntities), _1 = _0.next(); !_1.done; _1 = _0.next()) {
                        var entity = _1.value;
                        this.writebackRuleEntitiesMap[entity.id] = entity;
                    }
                }
                catch (e_8_1) {
                    e_8 = { error: e_8_1 };
                }
                finally {
                    try {
                        if (_1 && !_1.done && (_h = _0.return))
                            _h.call(_0);
                    }
                    finally {
                        if (e_8)
                            throw e_8.error;
                    }
                }
                try {
                    for (var _2 = __values(this.writebackProcessEntities), _3 = _2.next(); !_3.done; _3 = _2.next()) {
                        var entity = _3.value;
                        this.writebackProcessEntitiesMap[entity.id] = entity;
                    }
                }
                catch (e_9_1) {
                    e_9 = { error: e_9_1 };
                }
                finally {
                    try {
                        if (_3 && !_3.done && (_j = _2.return))
                            _j.call(_2);
                    }
                    finally {
                        if (e_9)
                            throw e_9.error;
                    }
                }
                try {
                    for (var _4 = __values(this.generalTextEntities), _5 = _4.next(); !_5.done; _5 = _4.next()) {
                        var entity = _5.value;
                        this.generalTextEntitiesMap[entity.id] = entity;
                    }
                }
                catch (e_10_1) {
                    e_10 = { error: e_10_1 };
                }
                finally {
                    try {
                        if (_5 && !_5.done && (_k = _4.return))
                            _k.call(_4);
                    }
                    finally {
                        if (e_10)
                            throw e_10.error;
                    }
                }
            };
        return ExportLogResult;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/ExpressionConditionContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ExpressionConditionContent = /** @class */ (function (_super) {
        __extends(ExpressionConditionContent, _super);
        function ExpressionConditionContent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        ExpressionConditionContent.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.expressionContent = null;
            };
        /**
         * @return {?}
         */
        ExpressionConditionContent.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "expressionContent", this.expressionContent);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        ExpressionConditionContent.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.expressionContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "expressionContent");
            };
        return ExpressionConditionContent;
    }(ConditionContent));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/ImplClassConditionContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ImplClassConditionContent = /** @class */ (function (_super) {
        __extends(ImplClassConditionContent, _super);
        function ImplClassConditionContent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        ImplClassConditionContent.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.implClassTypeName = null;
            };
        /**
         * @return {?}
         */
        ImplClassConditionContent.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "implClassTypeName", this.implClassTypeName);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        ImplClassConditionContent.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.implClassTypeName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "implClassTypeName");
            };
        return ImplClassConditionContent;
    }(ConditionContent));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/MappingComponentContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MappingComponentContent = /** @class */ (function (_super) {
        __extends(MappingComponentContent, _super);
        function MappingComponentContent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        MappingComponentContent.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.mappingConfig = null;
            };
        /**
         * @return {?}
         */
        MappingComponentContent.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConfig", this.mappingConfig);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        MappingComponentContent.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.mappingConfig = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConfig");
            };
        return MappingComponentContent;
    }(ComponentContent));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/MappingRuleContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MappingRuleContent = /** @class */ (function (_super) {
        __extends(MappingRuleContent, _super);
        function MappingRuleContent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        MappingRuleContent.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.mappingRuleMetadataId = null;
                this.mappingRuleId = null;
                this.mappingContent = null;
                this.mappingConfig = null;
            };
        /**
         * @return {?}
         */
        MappingRuleContent.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleMetadataId", this.mappingRuleMetadataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleId", this.mappingRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingContent", this.mappingContent);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConfig", this.mappingConfig);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        MappingRuleContent.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.mappingRuleMetadataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleMetadataId");
                this.mappingRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleId");
                this.mappingContent = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingContent");
                this.mappingConfig = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConfig");
            };
        return MappingRuleContent;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/RtBillCpAndWpEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RtBillCpAndWpEntity = /** @class */ (function (_super) {
        __extends(RtBillCpAndWpEntity, _super);
        function RtBillCpAndWpEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.isEnabled = false;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        RtBillCpAndWpEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.isEnabled = false;
            };
        /**
         * @return {?}
         */
        RtBillCpAndWpEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RtBillCpAndWpEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.isEnabled = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
            };
        return RtBillCpAndWpEntity;
    }(AbstractBillCpAndWpEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/VoActionConditionContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var VoActionConditionContent = /** @class */ (function (_super) {
        __extends(VoActionConditionContent, _super);
        function VoActionConditionContent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return VoActionConditionContent;
    }(ConditionContent));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/VoActionContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var VoActionContent = /** @class */ (function (_super) {
        __extends(VoActionContent, _super);
        function VoActionContent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return VoActionContent;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/AssociationType.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var AssociationType = {
        /**
         * 默认。
         */
        DEFAULT: "DEFAULT",
        /**
         * 生成关系。
         */
        GENERATE: "GENERATE",
        /**
         * 参照关系。
         */
        REFER: "REFER",
        /**
         * 核销关系。
         */
        VERIFICATION: "VERIFICATION",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/DocumentAssociationEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var DocumentAssociationEntity = /** @class */ (function (_super) {
        __extends(DocumentAssociationEntity, _super);
        function DocumentAssociationEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.associationType = AssociationType.DEFAULT;
            _this.sequenceNum = 0;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        DocumentAssociationEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.associationType = AssociationType.DEFAULT;
                this.sequenceNum = 0;
                this.originDocuSucode = null;
                this.originDocuVoId = null;
                this.originDocuTypeId = null;
                this.originDocuDataId = null;
                this.originDocuItemId = null;
                this.originDocuItemCode = null;
                this.originDocuDetailId = null;
                this.originDocuDataIdPath = null;
                this.upDocuSucode = null;
                this.upDocuVoId = null;
                this.upDocuTypeId = null;
                this.upDocuDataId = null;
                this.upDocuItemId = null;
                this.upDocuDetailId = null;
                this.upDocuDataIdPath = null;
                this.upDocuCode = null;
                this.upDocuDate = null;
                this.upDocuAbstract = null;
                this.downDocuSucode = null;
                this.downDocuVoId = null;
                this.downDocuTypeId = null;
                this.downDocuDataId = null;
                this.downDocuItemId = null;
                this.downDocuDetailId = null;
                this.downDocuDataIdPath = null;
                this.downDocuCode = null;
                this.downDocuDate = null;
                this.downDocuAbstract = null;
            };
        /**
         * @return {?}
         */
        DocumentAssociationEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "associationType", this.associationType);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sequenceNum", this.sequenceNum);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuSucode", this.originDocuSucode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuVoId", this.originDocuVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuTypeId", this.originDocuTypeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuDataId", this.originDocuDataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuItemId", this.originDocuItemId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuItemCode", this.originDocuItemCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuDetailId", this.originDocuDetailId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuDataIdPath", this.originDocuDataIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuSucode", this.upDocuSucode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuVoId", this.upDocuVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuTypeId", this.upDocuTypeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDataId", this.upDocuDataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuItemId", this.upDocuItemId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDetailId", this.upDocuDetailId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDataIdPath", this.upDocuDataIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuCode", this.upDocuCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDate", this.upDocuDate);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuAbstract", this.upDocuAbstract);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuSucode", this.downDocuSucode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuVoId", this.downDocuVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuTypeId", this.downDocuTypeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDataId", this.downDocuDataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuItemId", this.downDocuItemId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDetailId", this.downDocuDetailId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDataIdPath", this.downDocuDataIdPath);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuCode", this.downDocuCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDate", this.downDocuDate);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuAbstract", this.downDocuAbstract);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        DocumentAssociationEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.associationType = AssociationType[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "associationType")];
                this.sequenceNum = commonApi.BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "sequenceNum");
                this.originDocuSucode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuSucode");
                this.originDocuVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuVoId");
                this.originDocuTypeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuTypeId");
                this.originDocuDataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuDataId");
                this.originDocuItemId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuItemId");
                this.originDocuItemCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuItemCode");
                this.originDocuDetailId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuDetailId");
                this.originDocuDataIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuDataIdPath");
                this.upDocuSucode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuSucode");
                this.upDocuVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuVoId");
                this.upDocuTypeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuTypeId");
                this.upDocuDataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDataId");
                this.upDocuItemId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuItemId");
                this.upDocuDetailId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDetailId");
                this.upDocuDataIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDataIdPath");
                this.upDocuCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuCode");
                this.upDocuDate = commonApi.BpmnModelJsonHelper.GetAttributeValueAsDateTime(obj, "upDocuDate");
                this.upDocuAbstract = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuAbstract");
                this.downDocuSucode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuSucode");
                this.downDocuVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuVoId");
                this.downDocuTypeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuTypeId");
                this.downDocuDataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDataId");
                this.downDocuItemId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuItemId");
                this.downDocuDetailId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDetailId");
                this.downDocuDataIdPath = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDataIdPath");
                this.downDocuCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuCode");
                this.downDocuDate = commonApi.BpmnModelJsonHelper.GetAttributeValueAsDateTime(obj, "downDocuDate");
                this.downDocuAbstract = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuAbstract");
            };
        return DocumentAssociationEntity;
    }(commonApi.Entity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/AssociationFieldsType.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var AssociationFieldsType = {
        /**
         * 默认。
         */
        DEFAULT: "DEFAULT",
        /**
         * 主对主。
         */
        MAIN_MAIN: "MAIN_MAIN",
        /**
         * 主对从。
         */
        MAIN_ITEM: "MAIN_ITEM",
        /**
         * 主对从从。
         */
        MAIN_DETAIL: "MAIN_DETAIL",
        /**
         * 从对主。
         */
        ITEM_MAIN: "ITEM_MAIN",
        /**
         * 从对从。
         */
        ITEM_ITEM: "ITEM_ITEM",
        /**
         * 从对从从。
         */
        ITEM_DETAIL: "ITEM_DETAIL",
        /**
         * 从从对主。
         */
        DETAIL_MAIN: "DETAIL_MAIN",
        /**
         * 从从对从。
         */
        DETAIL_ITEM: "DETAIL_ITEM",
        /**
         * 从从对从从。
         */
        DETAIL_DETAIL: "DETAIL_DETAIL",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/FieldsAssociationEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FieldsAssociationEntity = /** @class */ (function (_super) {
        __extends(FieldsAssociationEntity, _super);
        function FieldsAssociationEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.sysInit = false;
            // upAssociationFieldsValueType: AssociationFieldsValueType = AssociationFieldsValueType.FIELD;
            // upDocuConstant: string;
            // downAssociationFieldsValueType: AssociationFieldsValueType = AssociationFieldsValueType.FIELD;
            // downDocuConstant: string;
            _this.associationFieldsType = AssociationFieldsType.DEFAULT;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        FieldsAssociationEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.code = null;
                this.name = null;
                this.sysInit = false;
                // this.docuRelationId = null;
                // this.upAssociationFieldsValueType = AssociationFieldsValueType.FIELD;
                // this.downAssociationFieldsValueType = AssociationFieldsValueType.FIELD;
                // this.upDocuConstant = null;
                // this.downDocuConstant = null;
                this.associationFieldsType = AssociationFieldsType.DEFAULT;
                this.upDocuMainCode = null;
                this.upDocuMainField = null;
                this.upDocuMainFieldName = null;
                this.downDocuMainCode = null;
                this.downDocuMainField = null;
                this.downDocuMainFieldName = null;
                this.upDocuItemCode = null;
                this.upDocuItemField = null;
                this.upDocuItemFieldName = null;
                this.downDocuItemCode = null;
                this.downDocuItemField = null;
                this.downDocuItemFieldName = null;
                this.upDocuDetailCode = null;
                this.upDocuDetailField = null;
                this.upDocuDetailFieldName = null;
                this.downDocuDetailCode = null;
                this.downDocuDetailField = null;
                this.downDocuDetailFieldName = null;
                this.mapUpMainField = null;
                this.mapUpMainFieldName = null;
                this.mapDownMainField = null;
                this.mapDownMainFieldName = null;
                this.mapUpItemCode = null;
                this.mapUpItemField = null;
                this.mapUpItemFieldName = null;
                this.mapDownItemCode = null;
                this.mapDownItemField = null;
                this.mapDownItemFieldName = null;
                this.mapUpDetailCode = null;
                this.mapUpDetailField = null;
                this.mapUpDetailFieldName = null;
                this.mapDownDetailCode = null;
                this.mapDownDetailField = null;
                this.mapDownDetailFieldName = null;
                this.colI18nValues = null;
                this.suCode = null;
            };
        /**
         * @return {?}
         */
        FieldsAssociationEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sysInit", this.sysInit);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "docuRelationId", this.docuRelationId);
                // BpmnModelJsonHelper.SetAttributeValue(obj, "upAssociationFieldsValueType", this.upAssociationFieldsValueType);
                // BpmnModelJsonHelper.SetAttributeValue(obj, "downAssociationFieldsValueType", this.downAssociationFieldsValueType);
                // BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuConstant", this.upDocuConstant);
                // BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuConstant", this.downDocuConstant);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "associationFieldsType", this.associationFieldsType);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuMainCode", this.upDocuMainCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuMainField", this.upDocuMainField);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuMainFieldName", this.upDocuMainFieldName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuMainCode", this.downDocuMainCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuMainField", this.downDocuMainField);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuMainFieldName", this.downDocuMainFieldName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuItemCode", this.upDocuItemCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuItemField", this.upDocuItemField);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuItemFieldName", this.upDocuItemFieldName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuItemCode", this.downDocuItemCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuItemField", this.downDocuItemField);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuItemFieldName", this.downDocuItemFieldName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDetailCode", this.upDocuDetailCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDetailField", this.upDocuDetailField);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDetailFieldName", this.upDocuDetailFieldName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDetailCode", this.downDocuDetailCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDetailField", this.downDocuDetailField);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDetailFieldName", this.downDocuDetailFieldName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpMainField", this.mapUpMainField);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpMainFieldName", this.mapUpMainFieldName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownMainField", this.mapDownMainField);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownMainFieldName", this.mapDownMainFieldName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpItemCode", this.mapUpItemCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpItemField", this.mapUpItemField);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpItemFieldName", this.mapUpItemFieldName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownItemCode", this.mapDownItemCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownItemField", this.mapDownItemField);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownItemFieldName", this.mapDownItemFieldName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpDetailCode", this.mapUpDetailCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpDetailField", this.mapUpDetailField);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpDetailFieldName", this.mapUpDetailFieldName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownDetailCode", this.mapDownDetailCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownDetailField", this.mapDownDetailField);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownDetailFieldName", this.mapDownDetailFieldName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "colI18nValues", this.colI18nValues);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        FieldsAssociationEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.code = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
                this.name = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
                this.sysInit = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "sysInit");
                this.docuRelationId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "docuRelationId");
                // const upAssociationFieldsValueType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upAssociationFieldsValueType");
                // if (upAssociationFieldsValueType != null && upAssociationFieldsValueType != "") {
                //   this.upAssociationFieldsValueType = AssociationFieldsValueType[upAssociationFieldsValueType];
                // }
                // const downAssociationFieldsValueType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downAssociationFieldsValueType");
                // if (downAssociationFieldsValueType != null && downAssociationFieldsValueType != "") {
                //   this.downAssociationFieldsValueType = AssociationFieldsValueType[downAssociationFieldsValueType];
                // }
                // this.upDocuConstant = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuConstant");
                // this.downDocuConstant = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuConstant");
                this.associationFieldsType = AssociationFieldsType[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "associationFieldsType")];
                this.upDocuMainCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuMainCode");
                this.upDocuMainField = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuMainField");
                this.upDocuMainFieldName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuMainFieldName");
                this.downDocuMainCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuMainCode");
                this.downDocuMainField = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuMainField");
                this.downDocuMainFieldName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuMainFieldName");
                this.upDocuItemCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuItemCode");
                this.upDocuItemField = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuItemField");
                this.upDocuItemFieldName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuItemFieldName");
                this.downDocuItemCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuItemCode");
                this.downDocuItemField = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuItemField");
                this.downDocuItemFieldName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuItemFieldName");
                this.upDocuDetailCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDetailCode");
                this.upDocuDetailField = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDetailField");
                this.upDocuDetailFieldName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDetailFieldName");
                this.downDocuDetailCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDetailCode");
                this.downDocuDetailField = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDetailField");
                this.downDocuDetailFieldName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDetailFieldName");
                this.mapUpMainField = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpMainField");
                this.mapUpMainFieldName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpMainFieldName");
                this.mapDownMainField = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownMainField");
                this.mapDownMainFieldName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownMainFieldName");
                this.mapUpItemCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpItemCode");
                this.mapUpItemField = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpItemField");
                this.mapUpItemFieldName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpItemFieldName");
                this.mapDownItemCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownItemCode");
                this.mapDownItemField = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownItemField");
                this.mapDownItemFieldName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownItemFieldName");
                this.mapUpDetailCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpDetailCode");
                this.mapUpDetailField = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpDetailField");
                this.mapUpDetailFieldName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpDetailFieldName");
                this.mapDownDetailCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownDetailCode");
                this.mapDownDetailField = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownDetailField");
                this.mapDownDetailFieldName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownDetailFieldName");
                this.colI18nValues = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "colI18nValues");
                this.suCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
            };
        return FieldsAssociationEntity;
    }(commonApi.Entity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/AssociationLocation.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var AssociationLocation = {
        /**
         * 默认。
         */
        DEFAULT: "DEFAULT",
        /**
         * 上游单据记录。
         */
        UpBill: "UpBill",
        /**
         * 下游单据记录。
         */
        DownBill: "DownBill",
        /**
         * 中间表记录。
         */
        ThirdTable: "ThirdTable",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/RelationType.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var RelationType = {
        /**
         * 默认。
         */
        DEFAULT: "DEFAULT",
        /**
         * 一对多。
         */
        ONETOONE: "ONETOONE",
        /**
         * 一对多。
         */
        ONETOMANY: "ONETOMANY",
        /**
         * 多对一。
         */
        MANYTOONE: "MORETOONE",
        /**
         * 多对多。
         */
        MANYTOMANY: "MANYTOMANY",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/TrackingMode.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var TrackingMode = {
        /**
         * 默认。
         */
        DEFAULT: "DEFAULT",
        /**
         * 单据流方式。
         */
        AIF: "AIF",
        /**
         * 设计时配置表方式。
         */
        DTRELATION: "DTRELATION",
        /**
         * 运行时业务数据表方式。
         */
        RTRELATION: "RTRELATION",
        /**
         * 自定义构件方式。
         */
        COMPONENT: "COMPONENT",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/TrackingType.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var TrackingType = {
        /**
         * 默认。
         */
        DEFAULT: "DEFAULT",
        /**
         * 生成关系追踪。
         */
        GENERATE: "GENERATE",
        /**
         * 业务自定义。
         */
        BIZCUSTOM: "BIZCUSTOM",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/DocumentRelationEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var DocumentRelationEntity = /** @class */ (function (_super) {
        __extends(DocumentRelationEntity, _super);
        function DocumentRelationEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.sysInit = false;
            _this.trackingType = TrackingType.DEFAULT;
            _this.trackingMode = TrackingMode.DEFAULT;
            _this.associationLocation = AssociationLocation.DEFAULT;
            _this.relationType = RelationType.DEFAULT;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        DocumentRelationEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.code = null;
                this.name = null;
                this.sysInit = false;
                this.upDocuTypeId = null;
                this.downDocuTypeId = null;
                this.trackingType = TrackingType.DEFAULT;
                this.trackingMode = TrackingMode.DEFAULT;
                this.associationLocation = AssociationLocation.DEFAULT;
                this.associationSuCode = null;
                this.associationSchemaId = null;
                this.associationDataSourceType = null;
                this.relationType = RelationType.DEFAULT;
                this.colI18nValues = null;
                this.trackingComponentId = null;
                this.upConditionExpressionText = null;
                this.downConditionExpressionText = null;
            };
        /**
         * @return {?}
         */
        DocumentRelationEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'code', this.code);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'name', this.name);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'sysInit', this.sysInit);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'upDocuTypeId', this.upDocuTypeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'downDocuTypeId', this.downDocuTypeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'trackingType', this.trackingType);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'trackingMode', this.trackingMode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'associationLocation', this.associationLocation);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'associationSuCode', this.associationSuCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'associationSchemaId', this.associationSchemaId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'associationDataSourceType', this.associationDataSourceType);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'relationType', this.relationType);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'colI18nValues', this.colI18nValues);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'trackingComponentId', this.trackingComponentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'upConditionExpressionText', this.upConditionExpressionText);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'downConditionExpressionText', this.downConditionExpressionText);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        DocumentRelationEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.code = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'code');
                this.name = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'name');
                this.sysInit = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, 'sysInit');
                this.upDocuTypeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'upDocuTypeId');
                this.downDocuTypeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'downDocuTypeId');
                this.trackingType = TrackingType[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'trackingType')];
                this.trackingMode = TrackingMode[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'trackingMode')];
                this.associationLocation = AssociationLocation[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'associationLocation')];
                this.associationSuCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'associationSuCode');
                this.associationSchemaId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'associationSchemaId');
                this.associationDataSourceType = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'associationDataSourceType');
                this.relationType = RelationType[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'relationType')];
                this.colI18nValues = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'colI18nValues');
                this.trackingComponentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'trackingComponentId');
                this.upConditionExpressionText = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'upConditionExpressionText');
                this.downConditionExpressionText = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'downConditionExpressionText');
            };
        return DocumentRelationEntity;
    }(commonApi.Entity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/DocumentTypeEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var DocumentTypeEntity = /** @class */ (function (_super) {
        __extends(DocumentTypeEntity, _super);
        function DocumentTypeEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.sysInit = false;
            // accessMode: AccessMode = AccessMode.DEFAULT;
            _this.jointQueryMode = JointQueryMode.DEFAULT;
            // dataSql: string;
            _this.extend = false;
            return _this;
        }
        /**
         * @return {?}
         */
        DocumentTypeEntity.prototype.DocumentTypeEntity = /**
         * @return {?}
         */
            function () {
            };
        /**
         * @protected
         * @return {?}
         */
        DocumentTypeEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.code = null;
                this.name = null;
                this.sysInit = false;
                this.moduleId = null;
                this.bizKind = null;
                this.bizType = null;
                this.suCode = null;
                this.schemaId = null;
                this.jointQueryMode = JointQueryMode.DEFAULT;
                this.jointQueryParameters = null;
                this.nodeName = null;
                this.customNodeName = null;
                this.nodeStyle = null;
                this.customNodeStyle = null;
                this.summaryFields = null;
                this.customSummaryFields = null;
                this.displayFields = null;
                this.customDisplayFields = null;
                this.extend = false;
                this.colI18nValues = null;
                this.rootId = null;
                this.parentId = null;
                this.enableConditionExpressionText = null;
                this.trackingSourceType = null;
                this.tableSchema = null;
                this.customTableSchema = null;
            };
        /**
         * @return {?}
         */
        DocumentTypeEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'code', this.code);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'name', this.name);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'sysInit', this.sysInit);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'moduleId', this.moduleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'bizKind', this.bizKind);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'bizType', this.bizType);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'suCode', this.suCode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'schemaId', this.schemaId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'jointQueryMode', this.jointQueryMode);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'jointQueryParameters', this.jointQueryParameters);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'nodeName', this.nodeName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'customNodeName', this.customNodeName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'nodeStyle', this.nodeStyle);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'customNodeStyle', this.customNodeStyle);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'summaryFields', this.summaryFields);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'customSummaryFields', this.customSummaryFields);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'displayFields', this.displayFields);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'customDisplayFields', this.customDisplayFields);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'extend', this.extend);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'colI18nValues', this.colI18nValues);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'rootId', this.rootId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'parentId', this.parentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'enableConditionExpressionText', this.enableConditionExpressionText);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'trackingSourceType', this.trackingSourceType);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'tableSchema', this.tableSchema);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, 'customTableSchema', this.customTableSchema);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        DocumentTypeEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.code = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'code');
                this.name = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'name');
                this.sysInit = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, 'sysInit');
                this.moduleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'moduleId');
                this.bizKind = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'bizKind');
                this.bizType = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'bizType');
                this.suCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'suCode');
                this.schemaId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'schemaId');
                this.jointQueryMode = JointQueryMode[commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'jointQueryMode')];
                this.jointQueryParameters = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'jointQueryParameters');
                this.nodeName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'nodeName');
                this.customNodeName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'customNodeName');
                this.nodeStyle = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'nodeStyle');
                this.customNodeStyle = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'customNodeStyle');
                this.summaryFields = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'summaryFields');
                this.customSummaryFields = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'customSummaryFields');
                this.displayFields = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'displayFields');
                this.customDisplayFields = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'customDisplayFields');
                this.extend = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, 'extend');
                this.colI18nValues = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'colI18nValues');
                this.rootId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'rootId');
                this.parentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'parentId');
                this.enableConditionExpressionText = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'enableConditionExpressionText');
                this.trackingSourceType = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'trackingSourceType');
                this.tableSchema = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'tableSchema');
                this.customTableSchema = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'customTableSchema');
            };
        return DocumentTypeEntity;
    }(commonApi.Entity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/ThirdSchemaInfoEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ThirdSchemaInfoEntity = /** @class */ (function (_super) {
        __extends(ThirdSchemaInfoEntity, _super);
        function ThirdSchemaInfoEntity() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        ThirdSchemaInfoEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.schemaCategoryId = null;
                this.schemaCategoryName = null;
                this.schemaConvertorClassTypeName = null;
                this.trackingConfigClassTypeName = null;
                this.suCode = null;
            };
        /**
         * @return {?}
         */
        ThirdSchemaInfoEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "schemaCategoryId", this.schemaCategoryId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "schemaCategoryName", this.schemaCategoryName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "schemaConvertorClassTypeName", this.schemaConvertorClassTypeName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "trackingConfigClassTypeName", this.trackingConfigClassTypeName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        ThirdSchemaInfoEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.schemaCategoryId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaCategoryId");
                this.schemaCategoryName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaCategoryName");
                this.schemaConvertorClassTypeName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaConvertorClassTypeName");
                this.trackingConfigClassTypeName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "trackingConfigClassTypeName");
                this.suCode = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
            };
        return ThirdSchemaInfoEntity;
    }(commonApi.Entity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/TableSchema.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var TableSchema = /** @class */ (function (_super) {
        __extends(TableSchema, _super);
        function TableSchema() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.maxDisplayNumber = 2000;
            _this.isMainTable = false;
            _this.isDisplay = true;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        TableSchema.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.code = null;
                this.maxDisplayNumber = 2000;
                this.isMainTable = false;
                this.isDisplay = true;
            };
        /**
         * @return {?}
         */
        TableSchema.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "maxDisplayNumber", this.maxDisplayNumber);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isMainTable", this.isMainTable);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isDisplay", this.isDisplay);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        TableSchema.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.code = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
                this.maxDisplayNumber = commonApi.BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "maxDisplayNumber");
                this.isMainTable = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isMainTable");
                this.isDisplay = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isDisplay");
            };
        return TableSchema;
    }(commonApi.JsonConvertableImpl));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/AssociationFieldsValueType.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var AssociationFieldsValueType = {
        /**
         * 常量方式
         */
        CONSTANT: "CONSTANT",
        /**
         * 字段方式
         */
        FIELD: "FIELD",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/impl/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/entity/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/slim-entity/impl/AbstractSlimBillProcessEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 简化单据流实体类。
     * @abstract
     */
    var /**
     * 简化单据流实体类。
     * @abstract
     */ AbstractSlimBillProcessEntity = /** @class */ (function (_super) {
        __extends(AbstractSlimBillProcessEntity, _super);
        function AbstractSlimBillProcessEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.sysInit = false;
            _this.enabledParentEventRouterCond = true;
            _this.enabledParentMappingCond = true;
            _this.origin = true;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        AbstractSlimBillProcessEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.code = null;
                this.upBillVoMetadataId = null;
                this.upBillVoId = null;
                this.downBillVoMetadataId = null;
                this.downBillVoId = null;
                this.bizEventId = null;
                this.isSupportAutomatinFlow = false;
                this.sysInit = false;
                this.parentId = null;
                this.enabledParentEventRouterCond = true;
                this.enabledParentMappingCond = true;
                this.upBillBeId = null;
                this.downBillBeId = null;
                this.origin = true;
            };
        /**
         * @return {?}
         */
        AbstractSlimBillProcessEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataId", this.upBillVoMetadataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoId", this.upBillVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataId", this.downBillVoMetadataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoId", this.downBillVoId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventId", this.bizEventId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportAutomatinFlow", this.isSupportAutomatinFlow);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "sysInit", this.sysInit);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "parentId", this.parentId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "enabledParentEventRouterCond", this.enabledParentEventRouterCond);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "enabledParentMappingCond", this.enabledParentMappingCond);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeId", this.upBillBeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeId", this.downBillBeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "origin", this.origin);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        AbstractSlimBillProcessEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.code = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
                this.upBillVoMetadataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataId");
                this.upBillVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoId");
                this.downBillVoMetadataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataId");
                this.downBillVoId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoId");
                this.bizEventId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventId");
                this.isSupportAutomatinFlow = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportAutomatinFlow");
                this.sysInit = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "sysInit");
                this.parentId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "parentId");
                this.enabledParentEventRouterCond = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "enabledParentEventRouterCond");
                this.enabledParentMappingCond = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "enabledParentMappingCond");
                this.upBillBeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeId");
                this.downBillBeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeId");
                this.origin = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "origin");
            };
        return AbstractSlimBillProcessEntity;
    }(commonApi.SlimEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/slim-entity/impl/AbstractRtSlimBillProcessEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 运行时简化单据流实体类。
     * @abstract
     */
    var /**
     * 运行时简化单据流实体类。
     * @abstract
     */ AbstractRtSlimBillProcessEntity = /** @class */ (function (_super) {
        __extends(AbstractRtSlimBillProcessEntity, _super);
        function AbstractRtSlimBillProcessEntity() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.isEnabled = false;
            return _this;
        }
        /**
         * @protected
         * @return {?}
         */
        AbstractRtSlimBillProcessEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                this.isEnabled = false;
                this.eventRouterConditionsId = null;
                this.mappingConditionsId = null;
                this.mappingModeId = null;
                this.saveModeId = null;
            };
        /**
         * @return {?}
         */
        AbstractRtSlimBillProcessEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsId", this.eventRouterConditionsId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsId", this.mappingConditionsId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeId", this.mappingModeId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeId", this.saveModeId);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        AbstractRtSlimBillProcessEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.isEnabled = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
                this.eventRouterConditionsId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsId");
                this.mappingConditionsId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsId");
                this.mappingModeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeId");
                this.saveModeId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeId");
            };
        return AbstractRtSlimBillProcessEntity;
    }(AbstractSlimBillProcessEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/slim-entity/impl/RtSlimBillCreationProcessEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 运行时简化生单流实体类。
     */
    var /**
     * 运行时简化生单流实体类。
     */ RtSlimBillCreationProcessEntity = /** @class */ (function (_super) {
        __extends(RtSlimBillCreationProcessEntity, _super);
        function RtSlimBillCreationProcessEntity() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @protected
         * @return {?}
         */
        RtSlimBillCreationProcessEntity.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.formSettingsId = null;
                this.creationRuleId = null;
                this.downBillVoProxyMetadataId = null;
                this.downBillVoProxyId = null;
                this.upBillTrackingSettingsId = null;
                this.downBillTrackingSettingsId = null;
                this.reverseWriteRuleId = null;
                this.taskCenterConfig4Push = null;
                this.taskCenterConfig4Pull = null;
                this.isSupportPull = false;
                this.isSupportPush = false;
            };
        /**
         * @return {?}
         */
        RtSlimBillCreationProcessEntity.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "formSettingsId", this.formSettingsId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataId", this.downBillVoProxyMetadataId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyId", this.downBillVoProxyId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettingsId", this.upBillTrackingSettingsId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettingsId", this.downBillTrackingSettingsId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleId", this.reverseWriteRuleId);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Push", this.taskCenterConfig4Push);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Pull", this.taskCenterConfig4Pull);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPull", this.isSupportPull);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPush", this.isSupportPush);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RtSlimBillCreationProcessEntity.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.formSettingsId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettingsId");
                this.creationRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
                this.downBillVoProxyMetadataId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataId");
                this.downBillVoProxyId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyId");
                this.upBillTrackingSettingsId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettingsId");
                this.downBillTrackingSettingsId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettingsId");
                this.reverseWriteRuleId = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleId");
                this.taskCenterConfig4Push = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Push");
                this.taskCenterConfig4Pull = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Pull");
                this.isSupportPull = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPull");
                this.isSupportPush = commonApi.BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPush");
            };
        return RtSlimBillCreationProcessEntity;
    }(AbstractRtSlimBillProcessEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/slim-entity/impl/RtSlimBillCreationProcessEntityEx.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 单据生单流简化实体扩展类。
     */
    var /**
     * 单据生单流简化实体扩展类。
     */ RtSlimBillCreationProcessEntityEx = /** @class */ (function (_super) {
        __extends(RtSlimBillCreationProcessEntityEx, _super);
        function RtSlimBillCreationProcessEntityEx() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @param {?=} entity
         * @return {?}
         */
        RtSlimBillCreationProcessEntityEx.prototype.RtSlimBillCreationProcessEntityEx = /**
         * @param {?=} entity
         * @return {?}
         */
            function (entity) {
                if (entity != null) {
                    this.id = entity.id;
                    this.code = entity.code;
                    this.upBillVoId = entity.upBillVoId;
                    this.upBillVoMetadataId = entity.upBillVoMetadataId;
                    this.downBillVoId = entity.downBillVoId;
                    this.downBillVoMetadataId = entity.downBillVoMetadataId;
                    this.downBillVoProxyMetadataId = entity.downBillVoProxyMetadataId;
                    this.downBillVoProxyId = entity.downBillVoProxyId;
                    this.eventRouterConditionsId = entity.eventRouterConditionsId;
                    this.mappingConditionsId = entity.mappingConditionsId;
                    this.formSettingsId = entity.formSettingsId;
                    this.bizEventId = entity.bizEventId;
                    this.creationRuleId = entity.creationRuleId;
                    this.isEnabled = entity.isEnabled;
                    this.isSupportAutomatinFlow = entity.isSupportAutomatinFlow;
                    this.name = entity.name;
                    this.description = entity.description;
                }
            };
        /**
         * @protected
         * @return {?}
         */
        RtSlimBillCreationProcessEntityEx.prototype.ResetBeforeLoadFrom = /**
         * @protected
         * @return {?}
         */
            function () {
                _super.prototype.ResetBeforeLoadFrom.call(this);
                this.name = null;
                this.description = null;
                this.upBillVoName = null;
                this.downBillVoName = null;
            };
        /**
         * @return {?}
         */
        RtSlimBillCreationProcessEntityEx.prototype.ConvertToJsonObject = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var obj = _super.prototype.ConvertToJsonObject.call(this);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "description", this.description);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoName", this.upBillVoName);
                commonApi.BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoName", this.downBillVoName);
                return obj;
            };
        /**
         * @param {?} obj
         * @return {?}
         */
        RtSlimBillCreationProcessEntityEx.prototype.LoadFromJsonObject = /**
         * @param {?} obj
         * @return {?}
         */
            function (obj) {
                _super.prototype.LoadFromJsonObject.call(this, obj);
                this.name = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
                this.description = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "description");
                this.upBillVoName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoName");
                this.downBillVoName = commonApi.BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoName");
            };
        return RtSlimBillCreationProcessEntityEx;
    }(RtSlimBillCreationProcessEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/slim-entity/impl/RtSlimBillWritebackProcessEntity.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 运行时简化单据回写流实体类。
     */
    var /**
     * 运行时简化单据回写流实体类。
     */ RtSlimBillWritebackProcessEntity = /** @class */ (function (_super) {
        __extends(RtSlimBillWritebackProcessEntity, _super);
        function RtSlimBillWritebackProcessEntity() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return RtSlimBillWritebackProcessEntity;
    }(AbstractRtSlimBillProcessEntity));

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/slim-entity/impl/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/slim-entity/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/front-entity/BillTrackerFormContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BillTrackerFormContent = /** @class */ (function () {
        function BillTrackerFormContent() {
        }
        return BillTrackerFormContent;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/front-entity/BillTrackerNodeShownStyle.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BillTrackerNodeShownStyle = /** @class */ (function () {
        function BillTrackerNodeShownStyle() {
        }
        return BillTrackerNodeShownStyle;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/front-entity/BillTrackerSchema.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BillTrackerSchema = /** @class */ (function () {
        function BillTrackerSchema() {
        }
        return BillTrackerSchema;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/front-entity/BillTrackerSchemaFieldContent.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BillTrackerSchemaFieldContent = /** @class */ (function () {
        function BillTrackerSchemaFieldContent() {
        }
        return BillTrackerSchemaFieldContent;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/front-entity/BillTrackerSchemaTable.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var fullPathPrefix = "#";
    var BillTrackerSchemaTable = /** @class */ (function () {
        function BillTrackerSchemaTable() {
        }
        /**
         * 自定义JSON序列化方法
         * @returns
         */
        /**
         * 自定义JSON序列化方法
         * @return {?}
         */
        BillTrackerSchemaTable.prototype.toJSON = /**
         * 自定义JSON序列化方法
         * @return {?}
         */
            function () {
                /** @type {?} */
                var fields = this.fields && this.fields.map(( /**
                 * @param {?} originF
                 * @return {?}
                 */function (originF) {
                    /** @type {?} */
                    var sf = new commonApi.BillSchemaField();
                    // sf.fullPath = originF.fullPath;
                    sf.fieldPath = originF.fieldPath;
                    sf.fieldName = originF.fieldName;
                    sf.caption = originF.caption;
                    sf.displayName = originF.displayName;
                    sf.order = originF.order;
                    sf.i18nCols = originF.i18nCols;
                    sf.formatInfo = originF.formatInfo;
                    return sf;
                }));
                /** @type {?} */
                var obj = {
                    code: this.code,
                    caption: this.caption,
                    displayName: this.displayName,
                    order: this.order,
                    show: this.show,
                    fields: fields,
                    i18nCols: this.i18nCols,
                    maxDisplayNumber: this.maxDisplayNumber,
                    parentTable: this.parentTable,
                };
                return obj;
            };
        return BillTrackerSchemaTable;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/front-entity/JointQueryParameter.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var JointQueryParameter = /** @class */ (function () {
        function JointQueryParameter() {
        }
        return JointQueryParameter;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/front-entity/JointQueryParameters.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var JointQueryParameters = /** @class */ (function () {
        function JointQueryParameters() {
            this.params = [];
        }
        return JointQueryParameters;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/front-entity/JointQueryParameterType.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var JointQueryParameterType = {
        Constant: "Constant",
        Field: "Field",
        Expression: "Expression",
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/front-entity/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/index.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: public-api.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: edp-bif-common-api.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.JsonConvertableImpl = commonApi.JsonConvertableImpl;
    exports.BpmnModelJsonHelper = commonApi.BpmnModelJsonHelper;
    exports.TextValue = commonApi.TextValue;
    exports.SlimEntity = commonApi.SlimEntity;
    exports.Entity = commonApi.Entity;
    exports.CallType = commonApi.CallType;
    exports.GeneralTextEntity = commonApi.GeneralTextEntity;
    exports.FormSettings = commonApi.FormSettings;
    exports.AbstractRtBillProcessContent = AbstractRtBillProcessContent;
    exports.RtBillCreationProcessContent = RtBillCreationProcessContent;
    exports.RtBillWritebackProcessContent = RtBillWritebackProcessContent;
    exports.BillTrackingDataShownMode = BillTrackingDataShownMode;
    exports.BillTrackingNodeShownMode = BillTrackingNodeShownMode;
    exports.ConditionType = ConditionType;
    exports.MappingMode = MappingMode;
    exports.SaveMode = SaveMode;
    exports.TaskCenterConfig = TaskCenterConfig;
    exports.TaskCenterConfig4Pull = TaskCenterConfig4Pull;
    exports.TaskCenterConfig4Push = TaskCenterConfig4Push;
    exports.AbstractBillCpAndWpEntity = AbstractBillCpAndWpEntity;
    exports.AbstractBillCreationProcessEntity = AbstractBillCreationProcessEntity;
    exports.AbstractBillProcessEntity = AbstractBillProcessEntity;
    exports.AbstractBillWritebackProcessEntity = AbstractBillWritebackProcessEntity;
    exports.BillTrackingConfigEntity = BillTrackingConfigEntity;
    exports.BillTrackingSettings = BillTrackingSettings;
    exports.ComponentConditionContent = ComponentConditionContent;
    exports.ComponentContent = ComponentContent;
    exports.ConditionContent = ConditionContent;
    exports.ConditionEntity = ConditionEntity;
    exports.DtBillCpAndWpEntity = DtBillCpAndWpEntity;
    exports.DtBillCreationProcessEntity = DtBillCreationProcessEntity;
    exports.DtBillWritebackProcessEntity = DtBillWritebackProcessEntity;
    exports.ExportLogResult = ExportLogResult;
    exports.ExpressionConditionContent = ExpressionConditionContent;
    exports.ImplClassConditionContent = ImplClassConditionContent;
    exports.MappingComponentContent = MappingComponentContent;
    exports.MappingModeEntity = MappingModeEntity;
    exports.MappingRuleContent = MappingRuleContent;
    exports.RtBillCpAndWpEntity = RtBillCpAndWpEntity;
    exports.RtBillCreationProcessEntity = RtBillCreationProcessEntity;
    exports.RtBillWritebackProcessEntity = RtBillWritebackProcessEntity;
    exports.SaveModeEntity = SaveModeEntity;
    exports.VoActionConditionContent = VoActionConditionContent;
    exports.VoActionContent = VoActionContent;
    exports.AssociationType = AssociationType;
    exports.DocumentAssociationEntity = DocumentAssociationEntity;
    exports.FieldsAssociationEntity = FieldsAssociationEntity;
    exports.DocumentRelationEntity = DocumentRelationEntity;
    exports.DocumentTypeEntity = DocumentTypeEntity;
    exports.AssociationFieldsType = AssociationFieldsType;
    exports.JointQueryMode = JointQueryMode;
    exports.RelationType = RelationType;
    exports.TrackingMode = TrackingMode;
    exports.TrackingType = TrackingType;
    exports.AssociationLocation = AssociationLocation;
    exports.ThirdSchemaInfoEntity = ThirdSchemaInfoEntity;
    exports.TableSchema = TableSchema;
    exports.AssociationFieldsValueType = AssociationFieldsValueType;
    exports.AbstractRtSlimBillProcessEntity = AbstractRtSlimBillProcessEntity;
    exports.AbstractSlimBillProcessEntity = AbstractSlimBillProcessEntity;
    exports.RtSlimBillCreationProcessEntity = RtSlimBillCreationProcessEntity;
    exports.RtSlimBillCreationProcessEntityEx = RtSlimBillCreationProcessEntityEx;
    exports.RtSlimBillWritebackProcessEntity = RtSlimBillWritebackProcessEntity;
    exports.BillTrackerFormContent = BillTrackerFormContent;
    exports.BillTrackerNodeShownStyle = BillTrackerNodeShownStyle;
    exports.BillTrackerSchema = BillTrackerSchema;
    exports.BillTrackerSchemaFieldContent = BillTrackerSchemaFieldContent;
    exports.fullPathPrefix = fullPathPrefix;
    exports.BillTrackerSchemaTable = BillTrackerSchemaTable;
    exports.JointQueryParameter = JointQueryParameter;
    exports.JointQueryParameters = JointQueryParameters;
    exports.JointQueryParameterType = JointQueryParameterType;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
