// @farris/ui-datatable/esm5/lib/table/cell-style.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { DataTableBodyComponent } from './datatable-body.component';
var CellStylerDirective = /** @class */ (function () {
    function CellStylerDirective(dtb, elRef, render) {
        this.dtb = dtb;
        this.elRef = elRef;
        this.render = render;
        this.options = null;
    }
    /**
     * @return {?}
     */
    CellStylerDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.dtb && this.options) {
            var _a = this.options, val = _a.val, row = _a.row, col = _a.col, index = _a.index;
            /** @type {?} */
            var _sty_1 = null;
            if (!col) {
                _sty_1 = this.dtb.renderRowStyle(row, index) || {};
            }
            else {
                /** @type {?} */
                var rowStyle = this.dtb.renderRowStyle(row, index) || {};
                /** @type {?} */
                var cellStyle = this.dtb.renderCellStyle(val, row, index, col) || {};
                _sty_1 = cellStyle;
                if (col.fixed === 'right') {
                    /** @type {?} */
                    var hasBackground = Object.keys(cellStyle).find((/**
                     * @param {?} n
                     * @return {?}
                     */
                    function (n) {
                        return ['background', 'background-color', 'backgroundcolor'].indexOf(n.toLowerCase()) > -1;
                    }));
                    if (hasBackground !== null) {
                        _sty_1 = Object.assign(cellStyle, rowStyle);
                    }
                }
            }
            if (_sty_1) {
                Object.keys(_sty_1).forEach((/**
                 * @param {?} n
                 * @return {?}
                 */
                function (n) {
                    _this.elRef.nativeElement.style.setProperty(n, _sty_1[n], 'important');
                }));
            }
        }
    };
    CellStylerDirective.decorators = [
        { type: Directive, args: [{ selector: '[cell-styler]' },] }
    ];
    /** @nocollapse */
    CellStylerDirective.ctorParameters = function () { return [
        { type: DataTableBodyComponent },
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    CellStylerDirective.propDecorators = {
        options: [{ type: Input, args: ['cell-styler',] }]
    };
    return CellStylerDirective;
}());
export { CellStylerDirective };
if (false) {
    /** @type {?} */
    CellStylerDirective.prototype.options;
    /**
     * @type {?}
     * @private
     */
    CellStylerDirective.prototype.dtb;
    /**
     * @type {?}
     * @private
     */
    CellStylerDirective.prototype.elRef;
    /**
     * @type {?}
     * @private
     */
    CellStylerDirective.prototype.render;
}
