// @farris/ui-datatable/esm5/lib/table/datatable-footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ContentChild, TemplateRef } from '@angular/core';
var DatatableFooterComponent = /** @class */ (function () {
    function DatatableFooterComponent() {
    }
    DatatableFooterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'datatable-footer',
                    template: "\n    <div class=\"table\"\n    [class.table-hover]=\"hover\">\n        <ng-container *ngIf=\"!tableFooter\">\n                <table class=\"table table-hover\">\n                        <colgroup>\n                            <col class=\"dt-checkbox-cell\" />\n                            <col *ngFor=\"let col of columns\" [style.width]=\"col.width + 'px'\" />\n                        </colgroup>\n                        <thead>\n                            <tr>\n                                <th class=\"dt-checkbox-cell\">\n                                    \n                                </th>\n                                <th  *ngFor=\"let col of columns\" [attr.align]=\"col.align\" >{{ col.title }}</th>\n                            </tr>\n                        </thead>\n                    </table>\n        </ng-container>\n</div>\n    "
                }] }
    ];
    /** @nocollapse */
    DatatableFooterComponent.ctorParameters = function () { return []; };
    DatatableFooterComponent.propDecorators = {
        hover: [{ type: Input }],
        columns: [{ type: Input }],
        tableFooter: [{ type: Input }, { type: ContentChild, args: [TemplateRef,] }]
    };
    return DatatableFooterComponent;
}());
export { DatatableFooterComponent };
if (false) {
    /** @type {?} */
    DatatableFooterComponent.prototype.hover;
    /** @type {?} */
    DatatableFooterComponent.prototype.columns;
    /** @type {?} */
    DatatableFooterComponent.prototype.tableFooter;
}
