// @gsp-lcm/dbo-selector/fesm2015/gsp-lcm-dbo-selector.js
import { Injectable, Component, Input, Output, EventEmitter, ViewChild, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { GridModule } from '@progress/kendo-angular-grid';
import { FormsModule } from '@angular/forms';
import { HttpClient, HttpClientModule } from '@angular/common/http';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DboHelpDevService {
    /**
     * @param {?} rest
     */
    constructor(rest) {
        this.rest = rest;
        this.databaseurlDev = '/api/dev/main/v1.0/database-objects';
    }
    /**
     * 数据库对象选择帮助界面（设计时）
     * @param {?} relativePath
     * @return {?}
     */
    getAllDatabaseObject(relativePath) {
        /** @type {?} */
        const url = this.databaseurlDev + '?relativePath=' + relativePath;
        return this.rest.get(url).pipe();
    }
    /**
     * 数据库对象详细信息选择帮助界面
     * @param {?} relativePath
     * @param {?} databaseobjectid
     * @return {?}
     */
    getCurrentDatabaseObject(relativePath, databaseobjectid) {
        /** @type {?} */
        const url = this.databaseurlDev + '/single?relativePath=' + relativePath
            + '&databaseobjectid=' + databaseobjectid;
        return this.rest.get(url).pipe();
    }
}
DboHelpDevService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DboHelpDevService.ctorParameters = () => [
    { type: HttpClient }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DboSelectComponent {
    /**
     * @param {?} dboHelpService
     */
    constructor(dboHelpService) {
        this.dboHelpService = dboHelpService;
        this.url = '';
        /**
         * 帮助标题
         */
        this.title = '帮助';
        /**
         * 输出事件
         */
        this.selectedAfterOk = new EventEmitter();
        this.modalClosed = new EventEmitter();
        // 模式
        this.selectableSettings = {
            checkboxOnly: true,
            mode: 'single',
            enabled: true
        };
    }
    /**
     * 输入属性变化后
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // 服务端获取数据
        this.dboHelpService.getAllDatabaseObject(this.url)
            .subscribe((/**
         * @param {?} result
         * @return {?}
         */
        result => {
            this.data = result;
            this.curSelectedKeys = [];
            if (this.initSelectedKeys) {
                this.initSelectedKeys.map((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => this.curSelectedKeys.push(item)));
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.log('获取数据源失败！');
            console.log(error);
        }));
    }
    /**
     * 帮助确定按钮
     * @return {?}
     */
    btnOk() {
        // 根据选中行获取实体数据
        /** @type {?} */
        const selectPosEntity = [];
        this.curSelectedKeys.forEach((/**
         * @param {?} dboID
         * @return {?}
         */
        dboID => {
            /** @type {?} */
            const selectDBO = this.data.find((/**
             * @param {?} item
             * @return {?}
             */
            item => item['id'] === dboID));
            selectPosEntity.push(selectDBO);
        }));
        if (!selectPosEntity) {
            alert('请选择数据库对象');
            return;
        }
        else {
            // 发射当前选中数据
            this.selectedAfterOk.next(selectPosEntity);
        }
        this.hideModal();
    }
    /**
     * @return {?}
     */
    hideModal() {
        this.modalClosed.emit();
    }
}
DboSelectComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-dbo-select',
                template: "<div class=\"modal-body\">\r\n<div class=\"templateBlock\">\r\n  <kendo-grid [class]=\"'flex-auto'\" [selectable]=\"selectableSettings\" [kendoGridSelectBy]=\"'id'\" [selectedKeys]=\"curSelectedKeys\"\r\n    [kendoGridBinding]=\"data\">\r\n    <kendo-grid-checkbox-column width=\"10%\"></kendo-grid-checkbox-column>\r\n    <kendo-grid-column field=\"code\" title=\"\u7F16\u53F7\" width=\"40%\"></kendo-grid-column>\r\n    <kendo-grid-column field=\"name\" title=\"\u540D\u79F0\" width=\"45%\"></kendo-grid-column>\r\n  </kendo-grid>\r\n  </div>\r\n</div>\r\n<ng-template #btns>\r\n<button type=\"button\" class=\"btn btn-default btn-primary\" (click)=\"btnOk()\">\u786E\u5B9A</button>\r\n<button type=\"button\" class=\"btn btn-default\" (click)=\"hideModal()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
DboSelectComponent.ctorParameters = () => [
    { type: DboHelpDevService }
];
DboSelectComponent.propDecorators = {
    url: [{ type: Input }],
    title: [{ type: Input }],
    isShow: [{ type: Input }],
    initSelectedKeys: [{ type: Input }],
    selectedAfterOk: [{ type: Output }],
    modalClosed: [{ type: Output }],
    modalFooter: [{ type: ViewChild, args: ['modalFooter',] }],
    btns: [{ type: ViewChild, args: ['btns',] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DboSelectModule {
}
DboSelectModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    GridModule,
                    FormsModule,
                    HttpClientModule
                ],
                providers: [DboHelpDevService],
                declarations: [DboSelectComponent],
                entryComponents: [DboSelectComponent],
                exports: [DboSelectComponent]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { DboHelpDevService, DboSelectComponent, DboSelectModule };
