// @gsp-wf/wf-process-model/esm5/lib/model/participants/old/ActivityExecutorParticipant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { Participant } from "./Participant";
var ActivityExecutorParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(ActivityExecutorParticipant, _super);
    function ActivityExecutorParticipant(model, ActivityDefId, ActivityDefName) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = ActivityExecutorParticipant.Clr_Type_ID;
        _this.ActivityDefId = ActivityDefId;
        _this.ActivityDefName = ActivityDefName;
        return _this;
    }
    /**
     * @return {?}
     */
    ActivityExecutorParticipant.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "ActivityDefId", this.ActivityDefId);
        BpmnModelHelper.SetAttributeValue(obj, "ActivityDefName", this.ActivityDefName);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ActivityExecutorParticipant.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.ActivityDefId = obj.ActivityDefId;
        this.ActivityDefName = obj.ActivityDefName;
    };
    /**
     * @return {?}
     */
    ActivityExecutorParticipant.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new ActivityExecutorParticipant(this.Model);
    };
    /**
     * @return {?}
     */
    ActivityExecutorParticipant.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var activityExecutorParticipant = new ActivityExecutorParticipant(this.Model);
        activityExecutorParticipant.Id = BpmnModelHelper.GenerateElementId();
        activityExecutorParticipant.ActivityDefId = this.ActivityDefId;
        activityExecutorParticipant.ActivityDefName = this.ActivityDefName;
        if (this.Condition) {
            activityExecutorParticipant.Condition = this.Condition.clone();
        }
        return activityExecutorParticipant;
    };
    ActivityExecutorParticipant.Clr_Type_ID = 'wf.ActivityExecutorParticipant';
    ActivityExecutorParticipant.CLR_TYPE_NAME = '活动办理人';
    return ActivityExecutorParticipant;
}(Participant));
export { ActivityExecutorParticipant };
if (false) {
    /** @type {?} */
    ActivityExecutorParticipant.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    ActivityExecutorParticipant.CLR_TYPE_NAME;
    /** @type {?} */
    ActivityExecutorParticipant.prototype.ClrTypeID;
    /** @type {?} */
    ActivityExecutorParticipant.prototype.ActivityDefId;
    /** @type {?} */
    ActivityExecutorParticipant.prototype.ActivityDefName;
}
BpmnModelHelper.RegisterElementType(new ActivityExecutorParticipant(new BpmnModel()));
