// @gsp-wf/wf-process-model/esm5/lib/model/participants/old/Participant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { HumanPerformer } from "@edp-pmf/bpmn-model";
import { Condition } from "../../extension/condition-expression/Condition";
/**
 * @abstract
 */
var /**
 * @abstract
 */
Participant = /** @class */ (function (_super) {
    tslib_1.__extends(Participant, _super);
    function Participant(model, id, name) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "wf.Participant";
        _this.MexTag = 'wf.Participant';
        _this.Id = id;
        _this.Name = name;
        return _this;
    }
    /**
     * @return {?}
     */
    Participant.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    Participant.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
    };
    /**
     * @protected
     * @return {?}
     */
    Participant.prototype.GetExtensionElements4ConvertTo = /**
     * @protected
     * @return {?}
     */
    function () {
        /** @type {?} */
        var list = _super.prototype.GetExtensionElements4ConvertTo.call(this);
        if (this.Condition) {
            this.Condition.MexTag = this.MexTag;
            list.push(this.Condition);
        }
        return list;
    };
    /**
     * @protected
     * @param {?} extensionElements
     * @return {?}
     */
    Participant.prototype.SetExtensionElements4LoadFrom = /**
     * @protected
     * @param {?} extensionElements
     * @return {?}
     */
    function (extensionElements) {
        var e_1, _a;
        _super.prototype.SetExtensionElements4LoadFrom.call(this, extensionElements);
        try {
            for (var extensionElements_1 = tslib_1.__values(extensionElements), extensionElements_1_1 = extensionElements_1.next(); !extensionElements_1_1.done; extensionElements_1_1 = extensionElements_1.next()) {
                var item = extensionElements_1_1.value;
                if (item instanceof Condition) {
                    /** @type {?} */
                    var condition = (/** @type {?} */ (item));
                    if (condition.MexTag === this.MexTag) {
                        this.Condition = condition;
                    }
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (extensionElements_1_1 && !extensionElements_1_1.done && (_a = extensionElements_1.return)) _a.call(extensionElements_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    return Participant;
}(HumanPerformer));
/**
 * @abstract
 */
export { Participant };
if (false) {
    /** @type {?} */
    Participant.prototype.ClrTypeID;
    /** @type {?} */
    Participant.prototype.MexTag;
    /** @type {?} */
    Participant.prototype.Condition;
}
