// @gsp-wf/wf-process-model/esm5/lib/model/participants/old/UserParticipant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { Participant } from "./Participant";
var UserParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(UserParticipant, _super);
    function UserParticipant(model, userId, name, orgType) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = UserParticipant.Clr_Type_ID;
        _this.UserId = userId;
        _this.Name = name;
        _this.OrgType = orgType ? orgType : 'Sys';
        return _this;
    }
    /**
     * @return {?}
     */
    UserParticipant.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new UserParticipant(this.Model);
    };
    /**
     * @return {?}
     */
    UserParticipant.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "UserId", this.UserId);
        BpmnModelHelper.SetAttributeValue(obj, "OrgId", this.OrgId);
        BpmnModelHelper.SetAttributeValue(obj, "OrgName", this.OrgName);
        BpmnModelHelper.SetAttributeValue(obj, "OrgType", this.OrgType);
        BpmnModelHelper.SetAttributeValue(obj, "Tag", this.Tag);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    UserParticipant.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.OrgId = obj.OrgId;
        this.OrgName = obj.OrgName;
        this.OrgType = obj.OrgType;
        this.UserId = obj.UserId;
        this.Tag = obj.Tag;
    };
    /**
     * @return {?}
     */
    UserParticipant.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var userParticipant = new UserParticipant(this.Model);
        userParticipant.Id = BpmnModelHelper.GenerateElementId();
        userParticipant.Name = this.Name;
        userParticipant.OrgId = this.OrgId;
        userParticipant.OrgName = this.OrgName;
        userParticipant.OrgType = this.OrgType;
        userParticipant.UserId = this.UserId;
        userParticipant.Tag = this.Tag;
        if (this.Condition) {
            userParticipant.Condition = this.Condition.clone();
        }
        return userParticipant;
    };
    UserParticipant.Clr_Type_ID = "wf.UserParticipant";
    UserParticipant.CLR_TYPE_NAME = '用户';
    return UserParticipant;
}(Participant));
export { UserParticipant };
if (false) {
    /** @type {?} */
    UserParticipant.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    UserParticipant.CLR_TYPE_NAME;
    /** @type {?} */
    UserParticipant.prototype.ClrTypeID;
    /** @type {?} */
    UserParticipant.prototype.Tag;
    /** @type {?} */
    UserParticipant.prototype.UserId;
    /** @type {?} */
    UserParticipant.prototype.OrgId;
    /** @type {?} */
    UserParticipant.prototype.OrgName;
    /** @type {?} */
    UserParticipant.prototype.OrgType;
}
BpmnModelHelper.RegisterElementType(new UserParticipant(new BpmnModel()));
