// @edp-pmf/bpmn-model/dist/esm/lib/model/omgdi/Plane.js
import { __extends } from "tslib";
import { BpmnModelHelper } from "../BpmnModelHelper";
import { ValidateHelper } from "../ValidateHelper";
import { Node } from "./Node";
var Plane = /** @class */ (function (_super) {
    __extends(Plane, _super);
    function Plane(model) {
        var _this = _super.call(this, model) || this;
        _this.diagramElementDic = {};
        return _this;
    }
    Plane.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.diagramElementDic = {};
    };
    Plane.prototype.AddDiagramElement = function (element) {
        this.CheckDiagramElement(element);
        this.diagramElementDic[element.Id] = element;
    };
    Plane.prototype.GetDiagramElement = function (elementId) {
        if (elementId == null)
            return null;
        if (this.diagramElementDic.hasOwnProperty(elementId))
            return this.diagramElementDic[elementId];
        return null;
    };
    Plane.prototype.GetDiagramElements = function () {
        var diagramElements = new Array();
        for (var item in this.diagramElementDic) {
            var diagramElement = this.diagramElementDic[item];
            diagramElements.push(diagramElement);
        }
        return diagramElements;
    };
    Plane.prototype.RemoveDiagramElement = function (elementId) {
        delete this.diagramElementDic[elementId];
    };
    Plane.prototype.CheckDiagramElement = function (element) {
        ValidateHelper.CheckHasIdElement(element);
    };
    Plane.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        var diagramElements = new Array();
        for (var item in this.diagramElementDic) {
            var element = this.diagramElementDic[item].ConvertToJsonObject();
            diagramElements.push(element);
        }
        if (diagramElements.length > 0) {
            BpmnModelHelper.SetAttributeValue(obj, "diagramElements", diagramElements);
        }
        return obj;
    };
    Plane.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        if (obj.hasOwnProperty("diagramElements") && Array.isArray(obj.diagramElements)) {
            for (var _i = 0, _a = obj.diagramElements; _i < _a.length; _i++) {
                var node = _a[_i];
                var diagramElement = BpmnModelHelper.GetElementType(node.ClrTypeID);
                if (diagramElement == null)
                    continue;
                diagramElement.Model = this.Model;
                diagramElement.LoadFromJsonObject(node);
                this.AddDiagramElement(diagramElement);
            }
        }
    };
    Plane.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
        for (var _i = 0, _a = this.GetDiagramElements(); _i < _a.length; _i++) {
            var item = _a[_i];
            item.Validate(errors);
        }
    };
    return Plane;
}(Node));
export { Plane };
