// @edp-bif/runtime-api/esm5/lib/bill-tracker/components/child-info-list/child-info-list.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/components/child-info-list/child-info-list.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ComponentFactoryResolver, Injector, Input, ViewContainerRef } from "@angular/core";
import { LoadingService } from "@farris/ui-loading";
import { MessagerService } from "@farris/ui-messager";
import { BsModalService } from "@farris/ui-modal";
import { NotifyService } from "@farris/ui-notify";
import { BillInfo } from "../../../entity/tracking/BillInfo";
import { BillTrackerLocalePipe } from "../../pipe/bill-tracker-locale.pipe";
import { BillTrackerDataInfoService } from "../../services/bill-tracker-data-info.service";
var ChildInfoListComponent = /** @class */ (function () {
    function ChildInfoListComponent(dataInfoService, notify, messager, loading, modal, localePipe, vcr, inj, cfr, cdr) {
        this.dataInfoService = dataInfoService;
        this.notify = notify;
        this.messager = messager;
        this.loading = loading;
        this.modal = modal;
        this.localePipe = localePipe;
        this.vcr = vcr;
        this.inj = inj;
        this.cfr = cfr;
        this.cdr = cdr;
    }
    /**
     * @return {?}
     */
    ChildInfoListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    ChildInfoListComponent.decorators = [
        { type: Component, args: [{
                    selector: "lib-bill-tracker-child-info-list",
                    template: "<ng-container *ngFor=\"let childTable of childSchemaTables\">\r\n    <lib-bill-tracker-child-info [schemaTable]=\"childTable\" [dataTable]=\"childDataTableMap[childTable.tableName]\"\r\n        [billInfo]=\"billInfo\">\r\n    </lib-bill-tracker-child-info>\r\n</ng-container>"
                }] }
    ];
    /** @nocollapse */
    ChildInfoListComponent.ctorParameters = function () { return [
        { type: BillTrackerDataInfoService },
        { type: NotifyService },
        { type: MessagerService },
        { type: LoadingService },
        { type: BsModalService },
        { type: BillTrackerLocalePipe },
        { type: ViewContainerRef },
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: ChangeDetectorRef }
    ]; };
    ChildInfoListComponent.propDecorators = {
        childSchemaTables: [{ type: Input, args: ["childSchemaTables",] }],
        childDataTableMap: [{ type: Input, args: ["childDataTableMap",] }],
        billInfo: [{ type: Input, args: ["billInfo",] }]
    };
    return ChildInfoListComponent;
}());
export { ChildInfoListComponent };
if (false) {
    /** @type {?} */
    ChildInfoListComponent.prototype.childSchemaTables;
    /** @type {?} */
    ChildInfoListComponent.prototype.childDataTableMap;
    /** @type {?} */
    ChildInfoListComponent.prototype.billInfo;
    /** @type {?} */
    ChildInfoListComponent.prototype.bsModalRef;
    /** @type {?} */
    ChildInfoListComponent.prototype.dataInfoService;
    /** @type {?} */
    ChildInfoListComponent.prototype.notify;
    /** @type {?} */
    ChildInfoListComponent.prototype.messager;
    /** @type {?} */
    ChildInfoListComponent.prototype.loading;
    /** @type {?} */
    ChildInfoListComponent.prototype.modal;
    /** @type {?} */
    ChildInfoListComponent.prototype.localePipe;
    /** @type {?} */
    ChildInfoListComponent.prototype.vcr;
    /** @type {?} */
    ChildInfoListComponent.prototype.inj;
    /** @type {?} */
    ChildInfoListComponent.prototype.cfr;
    /** @type {?} */
    ChildInfoListComponent.prototype.cdr;
}
