// @farris/ide-devkit/esm5/lib/components/navtab/navtab.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgbNavTabset, NgbNavTab, NgbNavTabContent, NgbNavTabTitle } from './navtab';
import { NgbNavTabsetConfig } from './navtab-config';
export { NgbNavTabset, NgbNavTab, NgbNavTabContent, NgbNavTabTitle } from './navtab';
export { NgbNavTabsetConfig } from './navtab-config';
/** @type {?} */
var NGB_TABSET_DIRECTIVES = [NgbNavTabset, NgbNavTab, NgbNavTabContent, NgbNavTabTitle];
var NgbNavTabsetModule = /** @class */ (function () {
    function NgbNavTabsetModule() {
    }
    /**
     * @return {?}
     */
    NgbNavTabsetModule.forRoot = /**
     * @return {?}
     */
    function () { return { ngModule: NgbNavTabsetModule, providers: [NgbNavTabsetConfig] }; };
    NgbNavTabsetModule.decorators = [
        { type: NgModule, args: [{ declarations: NGB_TABSET_DIRECTIVES, exports: NGB_TABSET_DIRECTIVES, imports: [CommonModule] },] }
    ];
    return NgbNavTabsetModule;
}());
export { NgbNavTabsetModule };
