// @farris/ui-filter-panel/esm5/lib/filter-panel.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FilterPanelComponent } from './filter-panel.component';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { FilterPanelService } from './filter-panel.service';
import { InputGroupModule } from '@farris/ui-input-group';
import { FarrisDatePickerModule } from '@farris/ui-datepicker';
import { DatalistModule } from '@farris/ui-datalist';
import { ComboListModule } from '@farris/ui-combo-list';
import { LookupModule } from '@farris/ui-lookup';
import { FarrisFormsModule } from '@farris/ui-forms';
import { NumberSpinnerModule } from '@farris/ui-number-spinner';
import { PerfectScrollbarModule } from '@farris/ui-perfect-scrollbar';
import { LocaleModule } from '@farris/ui-locale';
import { PanelControlTypeXPipe } from './panelControlType.pipe';
import { FilterConditionModule, QuerySolutionService } from '@farris/ui-filter-condition';
import { PanelfilterCheckboxComponent } from './components/checkboxcomponent/checkbox.component';
import { PanelDateComponent } from './components/datecomponent/date.component';
import { PanelDateRangeComponent } from './components/daterangecomponent/daterange.component';
import { PanelDropDownComponent } from './components/dropdowncomponent/dropdown.component';
import { PanelHelpComponent } from './components/helpcomponent.ts/help.component';
import { PanelNumberComponent } from './components/numbercomponent/number.component';
import { PanelInputComponent } from './components/inputcomponent/input.component';
import { PanelSearchComponent } from './components/searchcomponent/search.component';
import { PanelRadioComponent } from './components/radiocomponent/radio.component';
import { PanelInputGroupComponent } from './components/inputgroupcomponent/inputgroup.component';
import { FilterSafePipe } from './safe.pipe';
import { FilterHighlightSearchPipe } from './highlightSearch.pipe';
import { PanelSingleNumberComponent } from './components/singlenumbercomponent/singlenumber.component';
import { PanelFlexibleNumberComponent } from './components/flexiblenumbercomponent/flexiblenumber.component';
import { SwitchModule } from '@farris/ui-switch';
import { PanelFlexibleDateComponent } from './components/flexibledatecomponent/flexibledate.component';
var FilterPanelModule = /** @class */ (function () {
    function FilterPanelModule() {
    }
    /**
     * @return {?}
     */
    FilterPanelModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: FilterPanelModule,
            providers: [
                FilterPanelService
            ]
        };
    };
    FilterPanelModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        FilterPanelComponent,
                        PanelControlTypeXPipe,
                        FilterHighlightSearchPipe,
                        FilterSafePipe,
                        PanelDateComponent,
                        PanelNumberComponent,
                        PanelHelpComponent,
                        PanelInputComponent,
                        PanelDropDownComponent,
                        PanelSearchComponent,
                        PanelfilterCheckboxComponent,
                        PanelDateRangeComponent,
                        PanelRadioComponent,
                        PanelInputGroupComponent,
                        PanelSingleNumberComponent,
                        PanelFlexibleNumberComponent,
                        PanelFlexibleDateComponent
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        InputGroupModule,
                        FarrisDatePickerModule,
                        DatalistModule,
                        ComboListModule,
                        LookupModule,
                        FarrisFormsModule,
                        NumberSpinnerModule,
                        PerfectScrollbarModule,
                        LocaleModule.forRoot(),
                        SwitchModule.forRoot(),
                        FilterConditionModule
                    ],
                    providers: [
                        // FilterPanelService,
                        QuerySolutionService
                        // FilterConditionService
                    ],
                    exports: [
                        FilterPanelComponent,
                        PanelControlTypeXPipe,
                        FilterHighlightSearchPipe,
                        FilterSafePipe,
                        PanelDateComponent,
                        PanelNumberComponent,
                        PanelHelpComponent,
                        PanelInputComponent,
                        PanelDropDownComponent,
                        PanelSearchComponent,
                        PanelfilterCheckboxComponent,
                        PanelDateRangeComponent,
                        PanelRadioComponent,
                        PanelInputGroupComponent,
                        PanelSingleNumberComponent,
                        PanelFlexibleNumberComponent,
                        PanelFlexibleDateComponent
                    ],
                    entryComponents: [
                        FilterPanelComponent
                    ]
                },] }
    ];
    return FilterPanelModule;
}());
export { FilterPanelModule };
