// @farris/ui-filter-panel/esm5/lib/filter-panel.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ComponentFactoryResolver, Injector, ApplicationRef } from '@angular/core';
import { CommonUtils } from '@farris/ui-common';
import { FilterPanelComponent } from './filter-panel.component';
import * as i0 from "@angular/core";
var FilterPanelService = /** @class */ (function () {
    function FilterPanelService(cfr, injector, appRef) {
        this.cfr = cfr;
        this.injector = injector;
        this.appRef = appRef;
        // private instanceMap = new Map();
        this.commonUtils = null;
        this.commonUtils = this.injector.get(CommonUtils, null);
        if (!this.commonUtils) {
            this.commonUtils = new CommonUtils();
        }
    }
    /**
     * @param {?} option
     * @return {?}
     */
    FilterPanelService.prototype.showPanel = /**
     * @param {?} option
     * @return {?}
     */
    function (option) {
        /** @type {?} */
        var instanceID = option.localStorageKey;
        this.hidePanel();
        /** @type {?} */
        var containerFac = this.cfr.resolveComponentFactory(FilterPanelComponent);
        /** @type {?} */
        var filterPanelContainer = containerFac.create(this.injector);
        this.appRef.attachView(filterPanelContainer.hostView);
        document.querySelector('body').appendChild(filterPanelContainer.location.nativeElement);
        filterPanelContainer.instance.option = option;
        filterPanelContainer.instance.injector = this.injector;
        filterPanelContainer.instance.zIndex = this.commonUtils.getFloatingLayerIndex();
        // filterPanelContainer.instance.clear.subscribe((event)=>{
        //   this.clear();
        // });
        filterPanelContainer.changeDetectorRef.markForCheck();
        filterPanelContainer.changeDetectorRef.detectChanges();
        // if (!this.instanceMap.has(instanceID)) {
        //     this.instanceMap.set(instanceID, filterPanelContainer);
        // }
        this.filterPanelContainer = filterPanelContainer;
        return filterPanelContainer.instance;
    };
    /**
     * @return {?}
     */
    FilterPanelService.prototype.hidePanel = /**
     * @return {?}
     */
    function () {
        //  const filterPanelContainer = this.instanceMap.get(id);
        if (this.filterPanelContainer) {
            this.filterPanelContainer.instance.hide = true;
            this.filterPanelContainer.destroy();
            this.filterPanelContainer = null;
            // this.instanceMap.delete(id);
        }
    };
    FilterPanelService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    FilterPanelService.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: Injector },
        { type: ApplicationRef }
    ]; };
    /** @nocollapse */ FilterPanelService.ngInjectableDef = i0.defineInjectable({ factory: function FilterPanelService_Factory() { return new FilterPanelService(i0.inject(i0.ComponentFactoryResolver), i0.inject(i0.INJECTOR), i0.inject(i0.ApplicationRef)); }, token: FilterPanelService, providedIn: "root" });
    return FilterPanelService;
}());
export { FilterPanelService };
if (false) {
    /** @type {?} */
    FilterPanelService.prototype.filterPanelContainer;
    /**
     * @type {?}
     * @private
     */
    FilterPanelService.prototype.commonUtils;
    /**
     * @type {?}
     * @private
     */
    FilterPanelService.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    FilterPanelService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    FilterPanelService.prototype.appRef;
}
