// @edp-bif/runtime-api/esm2015/lib/bill-tracker/config/bill-tracker-cell-css.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/config/bill-tracker-cell-css.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 预置弹出菜单样式
 * @type {?}
 */
export const billTrackerPopupMenuCss = `
body div.mxPopupMenu {
  -webkit-box-shadow: 3px 3px 6px #C0C0C0;
  -moz-box-shadow: 3px 3px 6px #C0C0C0;
  box-shadow: 3px 3px 6px #C0C0C0;
  background: white;
  position: absolute;
  border: 3px solid #e7e7e7;
  padding: 3px;
}
body table.mxPopupMenu {
  border-collapse: collapse;
  margin: 0px;
}
body tr.mxPopupMenuItem {
  color: black;
  cursor: default;
}
body td.mxPopupMenuItem {
  padding: 6px 60px 6px 30px;
  font-family: Arial;
  font-size: 10pt;
}
body td.mxPopupMenuIcon {
  background-color: white;
  padding: 0px;
}
body tr.mxPopupMenuItemHover {
  background-color: #eeeeee;
  color: black;
}
table.mxPopupMenu hr {
  border-top: solid 1px #cccccc;
}
table.mxPopupMenu tr {
  font-size: 4pt;
}
`;
/**
 * 预置节点样式类
 * @type {?}
 */
export const billTrackerCellCss = `
.bif-cell-node {
    position: relative;
    height: 100%;
    font-weight: normal;
    font-family: PingFangSC-Regular;
    display:block;
    text-align: left;
}
.bif-cell-node-main {
    height: 100%;
    background: #FFFFFF;
    border: 1px solid rgba(230,233,240,1);
    box-shadow: 0px 2px 10px 0px rgba(3,18,51,0.08);
    border-radius: 8px;
    border-left-width: 4px;
    border-left-color: #2A87FF;
}
.bif-cell-node-current {
    position: absolute;
    right: 0;
    top: 0;
    width: 50px;
    height: 28px;
    background-image: linear-gradient(39deg, #F5A144 0%, #FFC527 100%);
    box-shadow: 0px 2px 6px 0px rgba(245,161,68,0.4);
    border-radius: 0px 8px 0px 64px;
    color: white;
    text-align: center;
}
.bif-cell-node-current-text {
    display: block;
    margin: auto;
    line-height: 28px;
}
.bif-cell-node-title {
    padding-left:19px;
    padding-right:12px;
    padding-top: 16px;
    padding-bottom: 6px;
}
.bif-cell-node-title-text {
    font-size: 16px;
    color: #2D2F33;
    letter-spacing: 0;
    line-height: 24px;
    font-weight: 400;
    white-space: normal;
    text-overflow: ellipsis;
    overflow: hidden;
    word-break: break-all;
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-line-clamp: 1;
}
.bif-cell-node-summary-list {

}
.bif-cell-node-summary {
    display: flex;
    padding-left: 19px;
    padding-right:12px;
    font-size: 13px;
    color: #5A5E66;
    line-height: 24px;
    font-weight: 400;
}
.bif-cell-node-summary-caption {
    letter-spacing: 0;
    margin-right: 4px;
    width: 65px;
    min-width: 65px;
    text-align: right;
}
.bif-cell-node-summary-caption-text {
    white-space: normal;
    text-overflow: ellipsis;
    overflow: hidden;
    word-break: break-all;
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-line-clamp: 1;
}
.bif-cell-node-summary-value {

}
.bif-cell-node-summary-value-text {
    font-family: PingFangSC-Regular,HelveticaNeue,Helvetica,Arial,sans-serif;
    color: rgba(0,0,0,0.85);
    text-align: left;
    white-space: normal;
    display: -webkit-box;
    -webkit-box-orient: vertical;
    overflow: hidden;
    -webkit-line-clamp: 1;
    word-break: break-all;
    overflow: hidden;
    text-overflow: ellipsis;
}
.bif-cell-node-title-link {
}
.bif-cell-node-title-link:hover {
    text-decoration: underline;
    cursor: pointer;
}
.bif-cell-node-title-link-color {
    color: #56A0FF;
}
.bif-cell-node-title-only {
    justify-content: center;
    height: 100%;
    display: flex;
    align-items: center;
    border: 1px solid rgb(193,195,199,1);
}
.bif-cell-node-title-only-text {
    font-size: xx-large;
}
.bif-cell-node-thumb {
    height: 100%;
    background: #FFFFFF;
}
.bif-cell-node-thumb-img-container {
    height: 80%;
}
.bif-cell-node-thumb-img {
    height: 100%;
    width: 100%;
}
.bif-cell-node-thumb-title {
    height: 20%;
    display: flex;
    justify-content: center;
    display: flex;
    align-items: center;
}
.bif-cell-node-thumb-title-text {
    font-size: x-large;
}
.bif-cell-node-btn-group {
    position: absolute;
    bottom: 16px;
    right: 16px;
}
.bif-cell-node-btn {
    width: 22px;
    height: 22px;
    margin-left: 8px;
    cursor: pointer;
}
.bif-cell-node .bif-cell-node-btn-group-control {
    display: none;
}
.bif-cell-node:hover .bif-cell-node-btn-group-control {
    display: unset;
}
`;
