// @farris/ui-datagrid/esm5/lib/scrollbar/scrollbar.interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/** @type {?} */
export var SCROLLBAR_CONFIG = new InjectionToken('PERFECT_SCROLLBAR_CONFIG');
var Geometry = /** @class */ (function () {
    function Geometry(x, y, w, h) {
        this.x = x;
        this.y = y;
        this.w = w;
        this.h = h;
    }
    return Geometry;
}());
export { Geometry };
if (false) {
    /** @type {?} */
    Geometry.prototype.x;
    /** @type {?} */
    Geometry.prototype.y;
    /** @type {?} */
    Geometry.prototype.w;
    /** @type {?} */
    Geometry.prototype.h;
}
var Position = /** @class */ (function () {
    function Position(x, y) {
        this.x = x;
        this.y = y;
    }
    return Position;
}());
export { Position };
if (false) {
    /** @type {?} */
    Position.prototype.x;
    /** @type {?} */
    Position.prototype.y;
}
/** @type {?} */
export var ScrollbarEvents = [
    'psScrollY',
    'psScrollX',
    'psScrollUp',
    'psScrollDown',
    'psScrollLeft',
    'psScrollRight',
    'psYReachEnd',
    'psYReachStart',
    'psXReachEnd',
    'psXReachStart'
];
/**
 * @record
 */
export function ScrollbarConfigInterface() { }
if (false) {
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.handlers;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.wheelSpeed;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.swipeEasing;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.suppressScrollX;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.suppressScrollY;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.wheelPropagation;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.useBothWheelAxes;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.scrollingThreshold;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.minScrollbarLength;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.maxScrollbarLength;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.scrollXMarginOffset;
    /** @type {?|undefined} */
    ScrollbarConfigInterface.prototype.scrollYMarginOffset;
}
var ScrollbarConfig = /** @class */ (function () {
    function ScrollbarConfig(config) {
        if (config === void 0) { config = {}; }
        this.assign(config);
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    ScrollbarConfig.prototype.assign = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        if (config === void 0) { config = {}; }
        for (var key in config) {
            this[(/** @type {?} */ (key))] = config[(/** @type {?} */ (key))];
        }
    };
    return ScrollbarConfig;
}());
export { ScrollbarConfig };
if (false) {
    /** @type {?} */
    ScrollbarConfig.prototype.handlers;
    /** @type {?} */
    ScrollbarConfig.prototype.wheelSpeed;
    /** @type {?} */
    ScrollbarConfig.prototype.swipeEasing;
    /** @type {?} */
    ScrollbarConfig.prototype.suppressScrollX;
    /** @type {?} */
    ScrollbarConfig.prototype.suppressScrollY;
    /** @type {?} */
    ScrollbarConfig.prototype.wheelPropagation;
    /** @type {?} */
    ScrollbarConfig.prototype.useBothWheelAxes;
    /** @type {?} */
    ScrollbarConfig.prototype.scrollingThreshold;
    /** @type {?} */
    ScrollbarConfig.prototype.minScrollbarLength;
    /** @type {?} */
    ScrollbarConfig.prototype.maxScrollbarLength;
    /** @type {?} */
    ScrollbarConfig.prototype.scrollXMarginOffset;
    /** @type {?} */
    ScrollbarConfig.prototype.scrollYMarginOffset;
}
