// @farris/mobile-devkit/esm5/lib/app/app_eventBus.js
import * as tslib_1 from "tslib";
import { Subject } from 'rxjs';
var AppEventBus = /** @class */ (function () {
    /**
     * 构造函数
     */
    function AppEventBus() {
        this.eventBus = new Subject();
        this.subscriptionsMap = new Map();
    }
    AppEventBus.prototype.triggerEvent = function (event) {
        this.eventBus.next(event);
    };
    AppEventBus.prototype.subscribe = function (componentId, func, funcError) {
        if (this.subscriptionsMap.has(componentId) === false) {
            var subscription = this.eventBus.subscribe(function (value) { return func(value); }, function (value) { return funcError(value); });
            this.subscriptionsMap.set(componentId, subscription);
        }
    };
    AppEventBus.prototype.unsubscribe = function (componentId) {
        if (this.subscriptionsMap.has(componentId) === true) {
            this.subscriptionsMap.get(componentId).unsubscribe();
            this.subscriptionsMap.delete(componentId);
        }
    };
    /**
     * 清除注册得指定key名监听器
     * @param keyStr 监听器key 默认值为listview上注册得
     * @returns
     */
    AppEventBus.prototype.unsubscribeAllListview = function (keyStr) {
        var e_1, _a;
        if (keyStr === void 0) { keyStr = '-listview-extend'; }
        if (this.subscriptionsMap.size == 0) {
            return;
        }
        try {
            // @ts-ignore: Unreachable code error
            for (var _b = tslib_1.__values(this.subscriptionsMap.keys()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var key = _c.value;
                if (key.indexOf(keyStr) > 0) {
                    this.subscriptionsMap.get(key).unsubscribe();
                    this.subscriptionsMap.delete(key);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    return AppEventBus;
}());
export { AppEventBus };
