// @farris/mobile-devkit/esm5/lib/app/view_model_contex_manager.js
/**
 * ViewModelContext管理类
 */
var ViewModelContextManager = /** @class */ (function () {
    /**
     * 构造函数
     */
    function ViewModelContextManager() {
        this.contextMap = new Map();
        this.contextSet = new Set();
    }
    /**
     * 注册Context
     */
    ViewModelContextManager.prototype.regContext = function (viewModelContext) {
        var id = viewModelContext.id;
        if (this.contextMap.has(id) === true) {
            // throw Error(`id为${id}的ViewModelContext已经存在`);
            // 临时处理，方式报错。
            this.unregContext(viewModelContext);
        }
        this.contextMap.set(id, viewModelContext);
        this.contextSet.add(viewModelContext);
    };
    /**
     * 取消注册
     */
    ViewModelContextManager.prototype.unregContext = function (context) {
        var id = context.id;
        this.contextMap.delete(id);
        this.contextSet.delete(context);
    };
    /**
     * 获取ContextMap
     */
    ViewModelContextManager.prototype.getContextMap = function () {
        return this.contextMap;
    };
    /**
     * 获取全部Context
     */
    ViewModelContextManager.prototype.getContexts = function () {
        return Array.from(this.contextSet);
    };
    /**
     * 根据命名控件获取上下文集合
     */
    ViewModelContextManager.prototype.getContextsByNamespace = function (namespace) {
        return this.getContexts();
    };
    /**
     * 根据id获取Context
     */
    ViewModelContextManager.prototype.getContextById = function (id) {
        var targetContext = this.contextMap.get(id);
        return targetContext;
    };
    /**
     * 获取根Context
     */
    ViewModelContextManager.prototype.getRootContext = function () {
        var contexts = this.getContexts();
        var rootContext = contexts.find(function (context) {
            return context.parent === null;
        });
        return rootContext;
    };
    /**
     * 获取传入视图模型id的root及root的下一代集合数组
     */
    ViewModelContextManager.prototype.getRootContextAndPosterityById = function (viewModelId) {
        var targetContext = this.getContextById(viewModelId);
        var contexts = this.getContexts();
        var contextsGroup = [];
        // 1.找到ROOT
        var RootId = this.getContextsGroupRoot(targetContext);
        // 2.可直接遍历拿到所有直系后代 (目前只有两层结构直接获取下一代)
        contexts.map(function (context) {
            if (context.parent && context.parent.id === RootId) {
                contextsGroup.push(context);
            }
        });
        contextsGroup.push(this.getContextById(RootId));
        return contextsGroup;
    };
    ViewModelContextManager.prototype.getContextsGroupRoot = function (context) {
        if (context.parent) {
            return this.getContextsGroupRoot(context.parent);
        }
        else {
            return context.id;
        }
    };
    return ViewModelContextManager;
}());
export { ViewModelContextManager };
