// @farris/ui-datagrid/esm5/lib/components/footer/datagrid-footer-cell.styler.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Renderer2, Injector, Input } from '@angular/core';
import { DatagridComponent } from '../../datagrid.component';
var FooterCellStylerDirective = /** @class */ (function () {
    function FooterCellStylerDirective(el, render, inject, dg) {
        this.el = el;
        this.render = render;
        this.inject = inject;
        this.dg = dg;
        this.isGroupFooter = false;
    }
    /**
     * @return {?}
     */
    FooterCellStylerDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    FooterCellStylerDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.isGroupFooter) {
            this.setGroupFooterCellStyle(this.column);
        }
        else {
            this.buildCustomCellStyle(this.column, this.rowData);
        }
    };
    /**
     * @private
     * @param {?} column
     * @param {?} rowData
     * @return {?}
     */
    FooterCellStylerDirective.prototype.buildCustomCellStyle = /**
     * @private
     * @param {?} column
     * @param {?} rowData
     * @return {?}
     */
    function (column, rowData) {
        /** @type {?} */
        var tr = this.el.nativeElement.parentNode;
        if (tr && !tr.id) {
            /** @type {?} */
            var rowdataid = this.dg.dfs.primaryId(rowData);
            /** @type {?} */
            var fixed = '';
            if (tr.closest('.f-datagrid-body-fixed-left')) {
                fixed = '-left';
            }
            else if (tr.closest('.f-datagrid-body-fixed-right')) {
                fixed = '-right';
            }
            /** @type {?} */
            var trID = 'footer-row' + fixed + '_' + (rowdataid || this.rowIndex);
            tr.id = trID;
        }
        if (!this.el.nativeElement.id) {
            this.el.nativeElement.id = 'footer-td_' + this.rowIndex + '_' + this.column.field.replace(/\./g, '_');
        }
        if (column && column.footer && column.footer.styler) {
            /** @type {?} */
            var _rd = rowData;
            /** @type {?} */
            var styler = column.footer.styler;
            /** @type {?} */
            var cs = styler(_rd[column.field], _rd);
            if (cs && Object.keys(cs).length) {
                this.dg.renderCustomStyle(cs, this.el.nativeElement);
            }
            else {
                this.dg.dgs.removeCssRule('#' + this.el.nativeElement.id);
            }
        }
    };
    /**
     * @private
     * @param {?} col
     * @return {?}
     */
    FooterCellStylerDirective.prototype.setGroupFooterCellStyle = /**
     * @private
     * @param {?} col
     * @return {?}
     */
    function (col) {
        var _this = this;
        // 分组合计行自定义样式 20210909
        if (col.groupFooter && col.groupFooter.styler) {
            /** @type {?} */
            var val = this.dg.commonUtils.getValue(col.field, this.rowData);
            /** @type {?} */
            var _customStyles_1 = this.dg.dgs.getCustomStyle(col.groupFooter.styler, [val, this.rowData]);
            if (_customStyles_1) {
                this.dg.renderCustomCls(_customStyles_1.cls, this.el.nativeElement);
                if (_customStyles_1.style) {
                    Object.keys(_customStyles_1.style).forEach((/**
                     * @param {?} n
                     * @return {?}
                     */
                    function (n) {
                        _this.el.nativeElement.style[n] = _customStyles_1.style[n];
                    }));
                }
            }
        }
    };
    FooterCellStylerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[footer-cell-styler]',
                },] }
    ];
    /** @nocollapse */
    FooterCellStylerDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: Injector },
        { type: DatagridComponent }
    ]; };
    FooterCellStylerDirective.propDecorators = {
        column: [{ type: Input }],
        rowData: [{ type: Input }],
        rowIndex: [{ type: Input }],
        isGroupFooter: [{ type: Input }]
    };
    return FooterCellStylerDirective;
}());
export { FooterCellStylerDirective };
if (false) {
    /** @type {?} */
    FooterCellStylerDirective.prototype.column;
    /** @type {?} */
    FooterCellStylerDirective.prototype.rowData;
    /** @type {?} */
    FooterCellStylerDirective.prototype.rowIndex;
    /** @type {?} */
    FooterCellStylerDirective.prototype.isGroupFooter;
    /**
     * @type {?}
     * @private
     */
    FooterCellStylerDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FooterCellStylerDirective.prototype.render;
    /**
     * @type {?}
     * @private
     */
    FooterCellStylerDirective.prototype.inject;
    /** @type {?} */
    FooterCellStylerDirective.prototype.dg;
}
