// @farris/ui-html-editor/esm2015/lib/image-resize-module/modules/DisplaySize.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BaseModule } from './BaseModule';
export class DisplaySize extends BaseModule {
    constructor() {
        super(...arguments);
        this.onCreate = (/**
         * @return {?}
         */
        () => {
            // Create the container to hold the size display
            this.display = document.createElement('div');
            // Apply styles
            Object.assign(this.display.style, this.options.displayStyles);
            // Attach it
            this.overlay.appendChild(this.display);
        });
        this.onDestroy = (/**
         * @return {?}
         */
        () => { });
        this.onUpdate = (/**
         * @return {?}
         */
        () => {
            if (!this.display || !this.img) {
                return;
            }
            /** @type {?} */
            const size = this.getCurrentSize();
            this.display.innerHTML = size.join(' &times; ');
            if (size[0] > 120 && size[1] > 30) {
                // position on top of image
                Object.assign(this.display.style, {
                    right: '4px',
                    bottom: '4px',
                    left: 'auto'
                });
            }
            else if (this.img.style.float == 'right') {
                // position off bottom left
                /** @type {?} */
                const dispRect = this.display.getBoundingClientRect();
                Object.assign(this.display.style, {
                    right: 'auto',
                    bottom: `-${dispRect.height + 4}px`,
                    left: `-${dispRect.width + 4}px`
                });
            }
            else {
                // position off bottom right
                /** @type {?} */
                const dispRect = this.display.getBoundingClientRect();
                Object.assign(this.display.style, {
                    right: `-${dispRect.width + 4}px`,
                    bottom: `-${dispRect.height + 4}px`,
                    left: 'auto'
                });
            }
        });
        this.getCurrentSize = (/**
         * @return {?}
         */
        () => [
            this.img.width,
            Math.round((this.img.width / this.img.naturalWidth) * this.img.naturalHeight)
        ]);
    }
}
if (false) {
    /** @type {?} */
    DisplaySize.prototype.display;
    /** @type {?} */
    DisplaySize.prototype.onCreate;
    /** @type {?} */
    DisplaySize.prototype.onDestroy;
    /** @type {?} */
    DisplaySize.prototype.onUpdate;
    /** @type {?} */
    DisplaySize.prototype.getCurrentSize;
}
