// @farris/ui-pagination/esm2015/lib/pagination.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FarrisCommonModule } from '@farris/ui-common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PaginatePipe } from './paginate.pipe';
import { PaginationService } from './pagination.service';
import { PaginationControlsComponent } from './pagination-controls.component';
import { PaginationControlsDirective } from './pagination-controls.directive';
import { LocaleModule } from '@farris/ui-locale';
import { PaginationSelectedListComponent } from './selected-list/selected-list.component';
export { PaginationService } from './pagination.service';
export { PaginationControlsComponent } from './pagination-controls.component';
export { PaginationControlsDirective } from './pagination-controls.directive';
export { PaginatePipe } from './paginate.pipe';
export class PaginationModule {
}
PaginationModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    FarrisCommonModule.forRoot(),
                    LocaleModule
                ],
                declarations: [
                    PaginatePipe,
                    PaginationControlsComponent,
                    PaginationControlsDirective,
                    PaginationSelectedListComponent
                ],
                providers: [PaginationService],
                exports: [
                    PaginatePipe,
                    PaginationControlsComponent,
                    PaginationControlsDirective,
                    PaginationSelectedListComponent
                ]
            },] }
];
