// @farris/ui-pagination/esm2015/lib/pagination.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { IdService } from '@farris/ui-common';
import { EventEmitter, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class PaginationService {
    constructor() {
        this.change = new EventEmitter();
        this.instances = {};
        this.DEFAULT_ID = 'DEFAULT_PAGINATION_ID';
    }
    /**
     * @return {?}
     */
    defaultId() {
        return this.DEFAULT_ID + new IdService().uuid();
    }
    /**
     * @param {?} id
     * @return {?}
     */
    removeInstance(id) {
        if (this.instances[id]) {
            this.instances[id] = null;
            delete this.instances[id];
        }
    }
    /**
     * @param {?} instance
     * @return {?}
     */
    register(instance) {
        if (instance.id == null) {
            instance.id = this.DEFAULT_ID;
        }
        if (!this.instances[instance.id]) {
            this.instances[instance.id] = instance;
            this.change.emit(instance.id);
        }
        else {
            /** @type {?} */
            const changed = this.updateInstance(instance);
            if (changed) {
                this.change.emit(instance.id);
            }
        }
    }
    /**
     * Check each property of the instance and update any that have changed. Return
     * true if any changes were made, else return false.
     * @private
     * @param {?} instance
     * @return {?}
     */
    updateInstance(instance) {
        /** @type {?} */
        let changed = false;
        for (const prop in this.instances[instance.id]) {
            if (instance[prop] !== this.instances[instance.id][prop]) {
                this.instances[instance.id][prop] = instance[prop];
                changed = true;
            }
        }
        return changed;
    }
    /**
     * Returns the current page number.
     * @param {?} id
     * @return {?}
     */
    getCurrentPage(id) {
        if (this.instances[id]) {
            return this.instances[id].currentPage;
        }
    }
    /**
     * Sets the current page number.
     * @param {?} id
     * @param {?} page
     * @return {?}
     */
    setCurrentPage(id, page) {
        if (this.instances[id]) {
            /** @type {?} */
            const instance = this.instances[id];
            /** @type {?} */
            const maxPage = Math.ceil(instance.totalItems / instance.itemsPerPage);
            if (page <= maxPage && 1 <= page) {
                this.instances[id].currentPage = page;
                this.change.emit(id);
            }
        }
    }
    /**
     * @param {?} id
     * @param {?} newList
     * @return {?}
     */
    setPageList(id, newList) {
        if (this.instances[id]) {
            /** @type {?} */
            const instance = this.instances[id];
            instance.pageList = newList;
            this.change.emit(id);
        }
    }
    /**
     * Sets the value of instance.totalItems
     * @param {?} id
     * @param {?} totalItems
     * @return {?}
     */
    setTotalItems(id, totalItems) {
        if (this.instances[id] && 0 <= totalItems) {
            this.instances[id].totalItems = totalItems;
            this.change.emit(id);
        }
    }
    /**
     * Sets the value of instance.itemsPerPage.
     * @param {?} id
     * @param {?} itemsPerPage
     * @return {?}
     */
    setItemsPerPage(id, itemsPerPage) {
        if (this.instances[id]) {
            this.instances[id].itemsPerPage = itemsPerPage;
            this.change.emit(id);
        }
    }
    /**
     * Returns a clone of the pagination instance object matching the id. If no
     * id specified, returns the instance corresponding to the default id.
     * @param {?=} id
     * @return {?}
     */
    getInstance(id = this.DEFAULT_ID) {
        if (this.instances[id]) {
            return this.clone(this.instances[id]);
        }
        return (/** @type {?} */ ({}));
    }
    /**
     * Perform a shallow clone of an object.
     * @private
     * @param {?} obj
     * @return {?}
     */
    clone(obj) {
        /** @type {?} */
        const target = {};
        for (const i in obj) {
            if (obj.hasOwnProperty(i)) {
                target[i] = obj[i];
            }
        }
        return target;
    }
}
PaginationService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ PaginationService.ngInjectableDef = i0.defineInjectable({ factory: function PaginationService_Factory() { return new PaginationService(); }, token: PaginationService, providedIn: "root" });
if (false) {
    /** @type {?} */
    PaginationService.prototype.change;
    /**
     * @type {?}
     * @private
     */
    PaginationService.prototype.instances;
    /**
     * @type {?}
     * @private
     */
    PaginationService.prototype.DEFAULT_ID;
}
