// @farris/ui-treetable/esm5/lib/high-light.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Pipe, ElementRef } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
var HighlightPipe = /** @class */ (function () {
    function HighlightPipe(sanitizer, el) {
        this.sanitizer = sanitizer;
        this.el = el;
    }
    /**
     * @param {?} text
     * @param {?} keyword
     * @param {?=} field
     * @param {?=} findField
     * @return {?}
     */
    HighlightPipe.prototype.transform = /**
     * @param {?} text
     * @param {?} keyword
     * @param {?=} field
     * @param {?=} findField
     * @return {?}
     */
    function (text, keyword, field, findField) {
        if (field === void 0) { field = ''; }
        if (findField === void 0) { findField = ''; }
        /** @type {?} */
        var needSearch = true;
        if (field && findField) {
            if (findField.indexOf(',') === -1) {
                needSearch = field === findField;
            }
            else {
                needSearch = findField.split(',').some((/**
                 * @param {?} n
                 * @return {?}
                 */
                function (n) { return n === field; }));
            }
        }
        if (!needSearch) {
            return text;
        }
        if (keyword && text) {
            /** @type {?} */
            var pattern = keyword.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, '\\$&');
            pattern = pattern.split(' ').filter((/**
             * @param {?} t
             * @return {?}
             */
            function (t) {
                return t.length > 0;
            })).join('|');
            /** @type {?} */
            var regex = new RegExp(pattern, 'gi');
            return this.sanitizer.bypassSecurityTrustHtml(
            // text.replace(regex, (match) => `<span class="search-highlight">${match}</span>`)
            this.replaceContent(text, regex, (/**
             * @param {?} match
             * @return {?}
             */
            function (match) { return "<span class=\"search-highlight\">" + match + "</span>"; })));
        }
        else {
            return text;
        }
    };
    /**
     * @param {?} html
     * @param {?} context
     * @return {?}
     */
    HighlightPipe.prototype.parseHTML = /**
     * @param {?} html
     * @param {?} context
     * @return {?}
     */
    function (html, context) {
        /** @type {?} */
        var t = (context || document).createElement('template');
        t.innerHTML = html;
        return t.content.cloneNode(true);
    };
    /**
     * @private
     * @param {?} strMatch1
     * @param {?} regex
     * @param {?} match
     * @return {?}
     */
    HighlightPipe.prototype.replaceContent = /**
     * @private
     * @param {?} strMatch1
     * @param {?} regex
     * @param {?} match
     * @return {?}
     */
    function (strMatch1, regex, match) {
        var e_1, _a;
        /** @type {?} */
        var matchReg1 = />.*?\</gi;
        // 先查出><之间的内容
        /** @type {?} */
        var strs = strMatch1.match(matchReg1);
        if (strs && strs.length) {
            try {
                for (var strs_1 = tslib_1.__values(strs), strs_1_1 = strs_1.next(); !strs_1_1.done; strs_1_1 = strs_1.next()) {
                    var aa = strs_1_1.value;
                    /** @type {?} */
                    var rep_old = strs[aa];
                    // 拿到原始符合字符串
                    /** @type {?} */
                    var rep_new = strs[aa].replace(regex, match);
                    // 替换成目的字符串
                    /** @type {?} */
                    var re2 = new RegExp(rep_old, 'g');
                    strMatch1 = strMatch1.replace(re2, rep_new); // 将符合的原始字符串替换成新字符串
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (strs_1_1 && !strs_1_1.done && (_a = strs_1.return)) _a.call(strs_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return strMatch1;
        }
        else {
            return strMatch1.replace(regex, match);
        }
    };
    HighlightPipe.decorators = [
        { type: Pipe, args: [{ name: 'highlight' },] }
    ];
    /** @nocollapse */
    HighlightPipe.ctorParameters = function () { return [
        { type: DomSanitizer },
        { type: ElementRef }
    ]; };
    return HighlightPipe;
}());
export { HighlightPipe };
if (false) {
    /** @type {?} */
    HighlightPipe.prototype.sanitizer;
    /**
     * @type {?}
     * @private
     */
    HighlightPipe.prototype.el;
}
