// @farris/ui-treetable/esm5/lib/hotkeys.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { EventManager } from '@angular/platform-browser';
import { Observable } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
/**
 * @record
 */
function HotkeyOptions() { }
if (false) {
    /** @type {?} */
    HotkeyOptions.prototype.element;
    /** @type {?} */
    HotkeyOptions.prototype.description;
    /** @type {?} */
    HotkeyOptions.prototype.keys;
}
var TTHotkeys = /** @class */ (function () {
    function TTHotkeys(eventManager) {
        this.eventManager = eventManager;
        this.hotkeys = new Map();
        this.defaults = {
            element: document
        };
    }
    /**
     * @param {?} options
     * @return {?}
     */
    TTHotkeys.prototype.addShortcut = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        var _this = this;
        /** @type {?} */
        var merged = tslib_1.__assign({}, this.defaults, options);
        if (merged.keys.indexOf('ctrl') > -1) {
            merged.keys = merged.keys.replace('ctrl', 'control');
        }
        /** @type {?} */
        var event = "keydown." + merged.keys;
        if (merged.description) {
            this.hotkeys.set(merged.keys, merged.description);
        }
        return new Observable((/**
         * @param {?} observer
         * @return {?}
         */
        function (observer) {
            /** @type {?} */
            var handler = (/**
             * @param {?} e
             * @return {?}
             */
            function (e) {
                e.preventDefault();
                observer.next(e);
            });
            /** @type {?} */
            var dispose = _this.eventManager.addEventListener(merged.element, event, handler);
            return (/**
             * @return {?}
             */
            function () {
                dispose();
                _this.hotkeys.delete(merged.keys);
            });
        }));
    };
    TTHotkeys.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TTHotkeys.ctorParameters = function () { return [
        { type: EventManager }
    ]; };
    /** @nocollapse */ TTHotkeys.ngInjectableDef = i0.defineInjectable({ factory: function TTHotkeys_Factory() { return new TTHotkeys(i0.inject(i1.EventManager)); }, token: TTHotkeys, providedIn: "root" });
    return TTHotkeys;
}());
export { TTHotkeys };
if (false) {
    /** @type {?} */
    TTHotkeys.prototype.hotkeys;
    /** @type {?} */
    TTHotkeys.prototype.defaults;
    /**
     * @type {?}
     * @private
     */
    TTHotkeys.prototype.eventManager;
}
