// @farris/ui-treetable/esm5/lib/virtualized-loader.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var VirtualizedLoaderService = /** @class */ (function () {
    function VirtualizedLoaderService(tt) {
        this.tt = tt;
    }
    /**
     * @return {?}
     */
    VirtualizedLoaderService.prototype.getTableHeight = /**
     * @return {?}
     */
    function () {
        return this.tt.height;
    };
    /**
     * @return {?}
     */
    VirtualizedLoaderService.prototype.getTableHeaderHeight = /**
     * @return {?}
     */
    function () {
        return this.tt.headerBox.nativeElement.getBoundingClientRect().height;
    };
    /**
     * @return {?}
     */
    VirtualizedLoaderService.prototype.getTableBodyHeight = /**
     * @return {?}
     */
    function () {
        return this.tt.height - this.getTableHeaderHeight();
    };
    /**
     * @return {?}
     */
    VirtualizedLoaderService.prototype.getTableWidth = /**
     * @return {?}
     */
    function () {
        return this.tt.width;
    };
    /**
     * @return {?}
     */
    VirtualizedLoaderService.prototype.getRowHeight = /**
     * @return {?}
     */
    function () {
        return this.tt.rowHeight;
    };
    /**
     * @param {?} scrollTop
     * @return {?}
     */
    VirtualizedLoaderService.prototype.getRowNodes = /**
     * @param {?} scrollTop
     * @return {?}
     */
    function (scrollTop) {
        /** @type {?} */
        var top = 0;
        /** @type {?} */
        var rows = [];
        /** @type {?} */
        var topHideHeight = 0;
        /** @type {?} */
        var bottomHideHeight = 0;
        /** @type {?} */
        var rowNodes = this.tt.getCurrentAllRowNodes();
        // const {rowNodes } = this.tt.state;
        /** @type {?} */
        var rowHeight = this.getRowHeight();
        /** @type {?} */
        var minTop = scrollTop;
        /** @type {?} */
        var maxTop = minTop + this.getTableHeight() + rowHeight;
        for (var i = 0; i < rowNodes.length; i++) {
            /** @type {?} */
            var n = rowNodes[i];
            if (!n.visible) {
                continue;
            }
            top += rowHeight;
            if (top + rowHeight < minTop) {
                topHideHeight += rowHeight;
                continue;
            }
            else {
                if (top > maxTop) {
                    bottomHideHeight += rowHeight;
                    continue;
                }
            }
            rows.push(n);
        }
        return {
            data: rows,
            topHideHeight: topHideHeight,
            bottomHideHeight: bottomHideHeight
        };
    };
    return VirtualizedLoaderService;
}());
export { VirtualizedLoaderService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    VirtualizedLoaderService.prototype.tt;
}
