// @gsp-cmp/ubcc-param-data-ui/esm5/lib/controls/param-control/number-control/number-control.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/controls/param-control/number-control/number-control.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var NumberControlComponent = /** @class */ (function () {
    function NumberControlComponent() {
        this.canEdit = true;
        this.precision = 0;
        this.max = NaN;
        this.min = NaN;
    }
    /**
     * @return {?}
     */
    NumberControlComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.param = this.param || {};
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    NumberControlComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.param = this.param || {};
        if (this.param && this.param.constraint) {
            this.max = Number(this.param.constraint.max) || NaN;
            this.min = Number(this.param.constraint.min) || NaN;
            this.precision = Number(this.param.constraint.precision) || 0;
        }
    };
    /**
     * @return {?}
     */
    NumberControlComponent.prototype.isReadonly = /**
     * @return {?}
     */
    function () {
        if (!this.canEdit) {
            return true;
        }
        else {
            return !this.param || !this.param.readonly ? false : true;
        }
    };
    /**
     * @return {?}
     */
    NumberControlComponent.prototype.getUnit = /**
     * @return {?}
     */
    function () {
        if (!this.param || !this.param.constraint || !this.param.constraint.unit) {
            return '';
        }
        else {
            return "(\u5355\u4F4D:" + this.param.constraint.unit + ")";
        }
    };
    /**
     * @return {?}
     */
    NumberControlComponent.prototype.show = /**
     * @return {?}
     */
    function () {
        return !this.param || !this.param.description ? false : true;
    };
    NumberControlComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-number-control',
                    template: "<div class=\"farris-form-controls-inline farris-group-wrap col-12\">\n  <div class=\"form-group farris-form-group\">\n    <label class=\"col-form-label col-4\">\n      <span class=\"farris-label-text\" style=\"float:right\">{{param.name}}{{getUnit()}}</span>\n    </label>\n    <div class=\"farris-input-wrap col-6\">\n      <farris-number-spinner [editable]=\"true\" [max]=\"max\" [min]=\"min\" [precision]=\"precision\" [readonly]=\"isReadonly()\" [(ngModel)]=\"param.value\" name=\"{{param.code}}\"></farris-number-spinner>\n    </div>\n    <span *ngIf=\"show()\" farrisTooltip [placement]=\"'right-top'\" [content]=\"param.description\" class=\"f-icon f-icon-info\"></span>\n  </div>\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    NumberControlComponent.ctorParameters = function () { return []; };
    NumberControlComponent.propDecorators = {
        param: [{ type: Input }],
        canEdit: [{ type: Input }]
    };
    return NumberControlComponent;
}());
export { NumberControlComponent };
if (false) {
    /** @type {?} */
    NumberControlComponent.prototype.param;
    /** @type {?} */
    NumberControlComponent.prototype.canEdit;
    /** @type {?} */
    NumberControlComponent.prototype.precision;
    /** @type {?} */
    NumberControlComponent.prototype.max;
    /** @type {?} */
    NumberControlComponent.prototype.min;
}
