// @gsp-wf/wf-dimension-def/esm2015/lib/service/wf-dimension-def.viewModel.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpService } from '@ecp-caf/caf-common';
import { Injectable } from '@angular/core';
export class WFDimensionDefViewmodel {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.url = '/api/runtime/wf/v1.0/dimensionDef';
    }
    /**
     * @param {?} flowFormId
     * @return {?}
     */
    loadDimensionDefs(flowFormId) {
        if (flowFormId) {
            /** @type {?} */
            const url = this.url + `?entryFormID=${flowFormId}`;
            return this.http.get(url);
        }
    }
    /**
     * @param {?} id
     * @return {?}
     */
    removeDimensionDef(id) {
        if (id) {
            /** @type {?} */
            const url = this.url + `/${id}`;
            return this.http.delete(url, {});
        }
    }
    /**
     * @param {?} dimensionDef
     * @return {?}
     */
    addDimensionDef(dimensionDef) {
        return this.http.post(this.url, dimensionDef);
    }
    /**
     * @param {?} dimensionDef
     * @return {?}
     */
    updateDimensionDef(dimensionDef) {
        return this.http.patch(this.url, dimensionDef);
    }
    /**
     * @param {?} flowFormId
     * @return {?}
     */
    getDimensionDefSet(flowFormId) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/dimensionDef/customSetting?entryFormId=${flowFormId}`;
        return this.http.get(url);
    }
}
WFDimensionDefViewmodel.decorators = [
    { type: Injectable }
];
/** @nocollapse */
WFDimensionDefViewmodel.ctorParameters = () => [
    { type: HttpService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    WFDimensionDefViewmodel.prototype.url;
    /**
     * @type {?}
     * @private
     */
    WFDimensionDefViewmodel.prototype.http;
}
