// @qdp/common/esm2015/lib/service/extend-service/data-renderextend.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @abstract
 */
export class DataRenderExtendService {
}
DataRenderExtendService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ DataRenderExtendService.ngInjectableDef = i0.defineInjectable({ factory: function DataRenderExtendService_Factory() { return new DataRenderExtendService(); }, token: DataRenderExtendService, providedIn: "root" });
if (false) {
    /**
     * @abstract
     * @param {?} $event
     * @return {?}
     */
    DataRenderExtendService.prototype.beforeQueryData = function ($event) { };
    /**
     * @abstract
     * @param {?} $event
     * @return {?}
     */
    DataRenderExtendService.prototype.beforeLoadData = function ($event) { };
    /**
     * @abstract
     * @param {?} $event
     * @return {?}
     */
    DataRenderExtendService.prototype.afterLoadData = function ($event) { };
}
