// @gsp-wf/wf-biz-navigation/esm2015/lib/wf-biz-navigation.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Injector, Output, EventEmitter, ViewChild, Input } from '@angular/core';
import { BillcatTreeComponent } from '@gsp-bef/billcat-tree';
import { HttpService } from '@ecp-caf/caf-common';
import { TranslatePipe } from './services/translate.pipe';
export class WfBizNavigationComponent {
    /**
     * @param {?} injector
     * @param {?} http
     * @param {?} translate
     */
    constructor(injector, http, translate) {
        this.injector = injector;
        this.http = http;
        this.translate = translate;
        this.selectNodeEmitter = new EventEmitter();
        this.bizObject = new EventEmitter();
        this.bizCategory = new EventEmitter();
        this.tab = 'billCategory';
        this.data = new Array();
        this.config = {
            expandLevel: 1
        };
        this.emptyData = new Array();
        this.emptyCols = [{
                field: 'code', width: 100, title: this.translate.transform('static.code'),
            }];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.getBillcategoriesByExtend();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        // this.billcat.treeConfig.expandLevel = 1;
    }
    /**
     * @param {?} args
     * @return {?}
     */
    tabChange(args) {
        this.tab = args.nextId;
        // console.log({ tabchange: args });
    }
    /**
     * @param {?} data
     * @return {?}
     */
    boTreeSelectionChanged(data) {
        if (data.data && data.data.isDetail === '1') {
            /** @type {?} */
            const node = {
                type: 'bo',
                nodeInfo: data
            };
            this.bizObject.emit(node);
            this.selectNodeEmitter.emit(node);
        }
    }
    /**
     * @param {?} data
     * @return {?}
     */
    billCategorySelectionChanged(data) {
        if (data && data.detail) {
            /** @type {?} */
            const node = {
                type: 'bill',
                nodeInfo: data
            };
            this.bizCategory.emit(node);
            this.selectNodeEmitter.emit(node);
        }
    }
    /**
     * @private
     * @return {?}
     */
    getBillcategoriesByExtend() {
        /** @type {?} */
        const url = '/api/runtime/bcc/v1.0/billcategory/billcategoriesByExtend?allextends=false&customconfig=processAssignManager';
        this.http.get(url).subscribe((/**
         * @param {?} re
         * @return {?}
         */
        (re) => {
            if (re) {
                this.data = re;
            }
        }));
    }
}
WfBizNavigationComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf-biz-navigation',
                template: "\r\n<farris-tabs #tabs [value]=\"tab\" [fill]=\"true\" [contentFill]=\"true\" (tabChange)=\"tabChange($event)\"\r\n    class=\"f-utils-fill h-100\" style=\"background: #fff;\">\r\n    <farris-tab [title]=\"'static.kind' | translate\" [id]=\"'billCategory'\" class=\"f-utils-fill h-100\"><!--title=\"\u4E1A\u52A1\u79CD\u7C7B\"-->\r\n        <ng-template #content>\r\n            <div *ngIf=\"data.length>0; else empty\" class=\"f-utils-flex-column h-100\">\r\n                <perfect-scrollbar [config]=\"{}\">\r\n                    <billcat-tree [data]=\"data\" [config]=\"config\"\r\n                        (selectNodeEmitter)=\"billCategorySelectionChanged($event)\">\r\n                    </billcat-tree>\r\n                </perfect-scrollbar>\r\n            </div>\r\n        </ng-template>\r\n    </farris-tab>\r\n    <farris-tab [title] = \"'static.object' | translate\" [id]=\"'bizObject'\" class=\"f-utils-fill h-100\"><!--title=\"\u4E1A\u52A1\u5BF9\u8C61\"-->\r\n        <ng-template #content>\r\n            <div class = \"f-utils-flex-column h-100\" style=\"position: relative;\">\r\n                <app-bo-tree-panel (selectionChanged)=\"boTreeSelectionChanged($event)\" [showIcon]=\"true\">\r\n                </app-bo-tree-panel>\r\n            </div>\r\n        </ng-template>\r\n    </farris-tab>\r\n</farris-tabs>\r\n\r\n<ng-template #empty>\r\n    <div class=\"f-utils-fill-flex-column h-100\">\r\n        <div class=\"f-datagrid-norecords\" style=\"height: 100%;\">\r\n            <div class=\"f-datagrid-norecords-content\" style=\"width: 310px;\">{{'static.noData' | translate}}</div>\r\n        </div>\r\n    </div>\r\n</ng-template>\r\n\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
WfBizNavigationComponent.ctorParameters = () => [
    { type: Injector },
    { type: HttpService },
    { type: TranslatePipe }
];
WfBizNavigationComponent.propDecorators = {
    selectNodeEmitter: [{ type: Output }],
    bizObject: [{ type: Output }],
    bizCategory: [{ type: Output }],
    tab: [{ type: Input }],
    billcat: [{ type: ViewChild, args: [BillcatTreeComponent,] }]
};
if (false) {
    /** @type {?} */
    WfBizNavigationComponent.prototype.selectNodeEmitter;
    /** @type {?} */
    WfBizNavigationComponent.prototype.bizObject;
    /** @type {?} */
    WfBizNavigationComponent.prototype.bizCategory;
    /** @type {?} */
    WfBizNavigationComponent.prototype.tab;
    /** @type {?} */
    WfBizNavigationComponent.prototype.billcat;
    /** @type {?} */
    WfBizNavigationComponent.prototype.data;
    /** @type {?} */
    WfBizNavigationComponent.prototype.config;
    /** @type {?} */
    WfBizNavigationComponent.prototype.emptyData;
    /** @type {?} */
    WfBizNavigationComponent.prototype.emptyCols;
    /**
     * @type {?}
     * @private
     */
    WfBizNavigationComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    WfBizNavigationComponent.prototype.http;
    /**
     * @type {?}
     * @private
     */
    WfBizNavigationComponent.prototype.translate;
}
