// @ecp-caf/common-structure/esm5/lib/convertor/property/property-convertor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DataTypeConvertUtil } from '../utils/data-type-convert-util';
import { ExtensionConvertor } from '../extension/extension-convertor';
var PropertyConvertor = /** @class */ (function () {
    function PropertyConvertor() {
    }
    /**
     * @param {?} obj
     * @param {?} property
     * @param {?} ctxt
     * @return {?}
     */
    PropertyConvertor.prototype.convertToObject = /**
     * @param {?} obj
     * @param {?} property
     * @param {?} ctxt
     * @return {?}
     */
    function (obj, property, ctxt) {
        if (!property) {
            return;
        }
        obj['kind'] = property.kind;
        obj['id'] = property.id;
        obj['code'] = property.code;
        obj['name'] = property.name;
        obj['description'] = property.description;
        obj['defaultValue'] = property.defaultValue;
        obj['required'] = (/** @type {?} */ (property.required));
        if (property.propertyType) {
            /** @type {?} */
            var kind = property.propertyType.getKind();
            if (!kind) {
                throw new Error('属性序列化时属性类型获取TypeKind为空格！');
            }
            obj['propertyType'] = DataTypeConvertUtil.convertToObject(property.propertyType, ctxt);
        }
        obj['constraint'] = null;
        if (property.constraint) {
            /** @type {?} */
            var kind = property.constraint.getKind();
            if (!kind) {
                throw new Error('属性序列化时属性约束获取ConstraintKind为空格！');
            }
            obj['constraint'] = ctxt.getConstraintSerializerFactory().get(kind).serialize(property.constraint, ctxt);
        }
        if (property.extension) {
            /** @type {?} */
            var extensionConvertor = new ExtensionConvertor();
            extensionConvertor.addToJObj(property, obj);
        }
    };
    /**
     * @param {?} property
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    PropertyConvertor.prototype.initFromObject = /**
     * @param {?} property
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    function (property, obj, ctxt) {
        if (!obj) {
            return;
        }
        property.kind = obj['kind'];
        property.id = obj['id'];
        property.code = obj['code'];
        property.name = obj['name'];
        property.defaultValue = obj['defaultValue'];
        property.required = (/** @type {?} */ (obj['required']));
        property.description = obj['description'];
        if (obj['propertyType']) {
            /** @type {?} */
            var kind = obj['propertyType']['kind'];
            if (!kind) {
                throw new Error('解析属性时数据类型的TypeKind为空！');
            }
            property.propertyType = DataTypeConvertUtil.initFromObject(obj['propertyType'], ctxt);
        }
        if (obj['constraint']) {
            /** @type {?} */
            var kind = obj['constraint']['kind'];
            if (!kind) {
                throw new Error('解析属性时约束类型的ConstraintKind为空！');
            }
            property.constraint = ctxt.getConstraintSerializerFactory().get(kind).deserialize(obj['constraint'], ctxt);
        }
        if (obj["extension"]) {
            /** @type {?} */
            var extensionConvertor = new ExtensionConvertor();
            extensionConvertor.addToObj(property, obj);
        }
    };
    return PropertyConvertor;
}());
export { PropertyConvertor };
