// @farris/ui-messager/fesm5/farris-ui-messager.js
import { Subject, of } from 'rxjs';
import { format } from 'date-fns';
import { Directive, Input, ElementRef, NgZone, InjectionToken, Component, Injector, HostBinding, ViewChild, Renderer2, Injectable, NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { CommonUtils, FarrisCommonModule } from '@farris/ui-common';
import { BsModalService, ModalModule } from '@farris/ui-modal';
import { LocaleService, LocaleModule } from '@farris/ui-locale';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var FarrisDefaultFocusDirective = /** @class */ (function () {
    function FarrisDefaultFocusDirective(el, ngZone) {
        this.el = el;
        this.ngZone = ngZone;
    }
    /**
     * @return {?}
     */
    FarrisDefaultFocusDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    FarrisDefaultFocusDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.button.defaultFocus) {
            this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () {
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    _this.el.nativeElement.focus();
                }));
            }));
        }
    };
    FarrisDefaultFocusDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-default-focus]'
                },] }
    ];
    /** @nocollapse */
    FarrisDefaultFocusDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: NgZone }
    ]; };
    FarrisDefaultFocusDirective.propDecorators = {
        button: [{ type: Input, args: ['farris-default-focus',] }]
    };
    return FarrisDefaultFocusDirective;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var MessagerDefaultConfig = {
    // width: 340,
    // height: 220,
    title: '系统提示'
};
/** @type {?} */
var MESSAGER_DEFAULT_CONFIG = new InjectionToken('Messager default options.');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MessagerComponent = /** @class */ (function () {
    function MessagerComponent(injector) {
        this.injector = injector;
        this.showFontSize = false;
        this._message = '';
        this.msg = '';
        this.cls = 'farris-messager';
        this.okText = '';
        this.cancelText = '';
        this.showOkButton = true;
        this.showCancelButton = true;
        this.readonly = false;
        this.enableWordCount = false;
        this.countType = 'length';
        this.exception = null;
        this.showLines = 4;
        this.fontSize = '18';
        this.inputType = 'textarea';
        this.placeholder = '';
        this.buttons = [];
        this.expand = false;
        this.needShowExpand = false;
        this.wordsTotal = '';
        this.wrodsTotalTips = '';
        this.utils = this.injector.get(CommonUtils);
        this.localeService = this.injector.get(LocaleService);
        this.render = this.injector.get(Renderer2);
    }
    Object.defineProperty(MessagerComponent.prototype, "message", {
        get: /**
         * @return {?}
         */
        function () {
            return this._message;
        },
        set: /**
         * @param {?} str
         * @return {?}
         */
        function (str) {
            this._message = (str === undefined || str === null) ? '' : '' + str;
            if (this.type === 'prompt' && this.utils) {
                this._message = this.utils.unescapeHtml(this._message);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerComponent.prototype, "fontsizeName", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue('messager.prompt.fontSize.name');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerComponent.prototype, "fontSmall", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue('messager.prompt.fontSize.small');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerComponent.prototype, "fontMiddle", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue('messager.prompt.fontSize.middle');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerComponent.prototype, "fontBig", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue('messager.prompt.fontSize.big');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerComponent.prototype, "fontLarge", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue('messager.prompt.fontSize.large');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerComponent.prototype, "fontHuge", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue('messager.prompt.fontSize.huge');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerComponent.prototype, "expandText", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue('messager.exception.expand');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerComponent.prototype, "collapseText", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue('messager.exception.collapse');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerComponent.prototype, "happend", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue('messager.exception.happend');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerComponent.prototype, "detail", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue('messager.exception.detail');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerComponent.prototype, "exceptionMsgMaxHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return window.innerHeight - 10 - 10 - 100 - 100 - 16;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    MessagerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.enableWordCount && _this.promptText) {
                _this.onTextChange(null);
            }
        }));
        this.type = this.type === 'exception' ? 'error' : this.type;
    };
    Object.defineProperty(MessagerComponent.prototype, "icon", {
        get: /**
         * @return {?}
         */
        function () {
            return 'f-icon-' + (this.type === 'exception' ? 'error' : this.type);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    MessagerComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.exception && _this.exception['message']) {
                /** @type {?} */
                var realWidth = _this.excepMsgEl.nativeElement.getBoundingClientRect().width.toFixed(2) + 'px';
                _this.render.setStyle(_this.allexcepEl.nativeElement, 'width', realWidth);
                _this.needShowExpand = _this.allexcepEl.nativeElement.offsetHeight > _this.excepMsgEl.nativeElement.offsetHeight;
            }
            else {
                _this.needShowExpand = false;
            }
        }), 10);
    };
    /**
     * @param {?} val
     * @param {?} textAreaEl
     * @return {?}
     */
    MessagerComponent.prototype.onFontSizeChange = /**
     * @param {?} val
     * @param {?} textAreaEl
     * @return {?}
     */
    function (val, textAreaEl) {
        // this.promptText.nativeElement
        // textAreaEl.nativeElement.style['font-size'] = val;
    };
    /**
     * @private
     * @return {?}
     */
    MessagerComponent.prototype.getWordsTotal = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var c = this.countType === 'length' ? this.promptText.nativeElement.value.length
            : this.maxLength - this.promptText.nativeElement.value.length;
        return c;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    MessagerComponent.prototype.onTextChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.wordsTotal = this.getWordsTotal();
        this.wrodsTotalTips = this.localeService.getValue('messager.prompt.tips.' + this.countType).replace('{0}', this.wordsTotal);
    };
    /**
     * @param {?} t
     * @param {?} $event
     * @return {?}
     */
    MessagerComponent.prototype.expandMsg = /**
     * @param {?} t
     * @param {?} $event
     * @return {?}
     */
    function (t, $event) {
        $event.stopPropagation();
        if (t) {
            this.render.setStyle(this.excepMsgEl.nativeElement, 'overflow', 'auto');
            this.render.removeStyle(this.excepMsgEl.nativeElement, '-webkit-line-clamp');
        }
        else {
            this.render.setStyle(this.excepMsgEl.nativeElement, 'overflow', 'hidden');
            this.render.setStyle(this.excepMsgEl.nativeElement, '-webkit-line-clamp', this.showLines);
        }
        this.expand = t;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    MessagerComponent.prototype.onCopyClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        $event.stopPropagation();
        if (this.exception && this.exception.copyButton) {
            this.exception.copyButton.onClick({ event: $event, cmpRef: this });
        }
    };
    /**
     * @param {?} msg
     * @param {?=} timeout
     * @return {?}
     */
    MessagerComponent.prototype.showMiniNotify = /**
     * @param {?} msg
     * @param {?=} timeout
     * @return {?}
     */
    function (msg, timeout) {
        var _this = this;
        if (timeout === void 0) { timeout = 700; }
        if (this.miniNotifyEl) {
            this.miniNotifyEl.nativeElement.innerText = msg;
            this.render.setStyle(this.miniNotifyEl.nativeElement, 'display', '');
            this.render.setStyle(this.miniNotifyEl.nativeElement, 'opacity', 0.8);
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.render.setStyle(_this.miniNotifyEl.nativeElement, 'opacity', 0);
                _this.render.setStyle(_this.miniNotifyEl.nativeElement, 'display', 'none');
            }), timeout);
        }
    };
    MessagerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-messager',
                    template: "<!-- <section class=\"modal-tips d-flex flex-row\" *ngIf=\"type!=='prompt'\">\r\n    <div class=\"modal-tips-iconwrap\">\r\n        <span class=\"f-icon\" [ngClass]=\"icon\"></span>\r\n    </div>\r\n    <div class=\"modal-tips-content flex-fill fixdiv\">\r\n        <p class=\"toast-msg\" [innerHTML]=\"message\"></p>\r\n    </div>\r\n</section> -->\r\n\r\n<section class=\"modal-tips d-flex flex-row\" *ngIf=\"type!=='prompt'\" [ngClass]=\"'messager-type-'+type\">\r\n    <div class=\"float-left modal-tips-iconwrap\">\r\n        <span class=\"f-icon\" [ngClass]=\" icon\"></span>\r\n    </div>\r\n    <div class=\"modal-tips-content\">\r\n        <p class=\"toast-msg-title\" [innerHTML]=\"message | safe: 'html'\"></p>\r\n        <p class=\"toast-msg-detail\" [innerHTML]=\"msg | safe: 'html'\" *ngIf=\"msg !=='' && !exception\"></p>\r\n        <div class=\"toast-msg-detail\" *ngIf=\"exception\">\r\n            <div *ngIf=\"exception?.date\">{{happend}}: {{exception?.date}}</div>\r\n            <div id=\"exception_error_msg\" #excepMsg *ngIf=\"exception?.message\"\r\n                style=\"overflow: hidden;text-overflow: ellipsis;display: -webkit-box;-webkit-box-orient: vertical;\"\r\n                [style.-webkit-line-clamp]=\"showLines\" [style.max-height.px]=\"exceptionMsgMaxHeight\">\r\n                {{detail}}: <span [innerHTML]=\"exception?.message | safe: 'html'\"></span>\r\n            </div>\r\n            <span style=\"display: block; text-align: right;color:#2A87FF;\" *ngIf=\"needShowExpand\">\r\n                <span (click)=\"expandMsg(true, $event)\" style=\"cursor: pointer;\" *ngIf=\"!expand\">{{expandText}}</span>\r\n                <span (click)=\"expandMsg(false,  $event)\" style=\"cursor: pointer;\" *ngIf=\"expand\">{{collapseText}}</span>\r\n            </span>\r\n        </div>\r\n    </div>\r\n</section>\r\n<section class=\"modal-tips d-flex flex-row\"  *ngIf=\"type==='prompt'\">\r\n    <div class=\"flex-fill fixdiv mb-1 mr-3\" *ngIf=\"inputType === 'textarea'\">\r\n        <textarea #promptText name=\"promptMessage\" class=\"form-control\" [style.fontSize.px]=\"fontSize\"  style=\"height: 100%;resize: none\"\r\n            id=\"exampleFormControlTextarea1\" rows=\"4\" [attr.maxLength]=\"maxLength\" (input)=\"onTextChange($event)\" \r\n            >{{ message }}</textarea>\r\n        \r\n    </div>\r\n    <div class=\"flex-fill fixdiv mr-4 d-flex\" style=\"align-items: center;\" *ngIf=\"inputType !== 'textarea'\">\r\n        <input #promptText type=\"text\" class=\"form-control\" [attr.maxLength]=\"maxLength\"  (input)=\"onTextChange($event)\" [value]=\"message\"\r\n        placeholder=\"{{ placeholder }}\">\r\n    </div>\r\n\r\n</section>\r\n<span #wordCount class=\"textarea-wordcount\" *ngIf=\"enableWordCount && maxLength && maxLength > 0\" [attr.title]=\"wrodsTotalTips\"\r\n    style=\"position: relative;bottom: 10px;right: 20px;cursor: pointer;text-align: right;\">\r\n    {{ wordsTotal  + ' / ' + maxLength }}\r\n</span>\r\n\r\n<div class=\"modal-footer\" *ngIf=\"!(buttons && buttons.length) && (okText || cancelText)\">\r\n    <div class=\"d-flex flex-fill textarea-font-size\" style=\"text-align: left\" *ngIf=\"type==='prompt' && showFontSize\">\r\n        <span style=\"width:75px\">{{fontsizeName}}</span>\r\n        <select #fontSizeEl class=\"form-control\" style=\"max-width:100px\" [(ngModel)]=\"fontSize\" (change)=\"onFontSizeChange($event.target.value, promptText)\">\r\n            <option value=\"14\">{{fontSmall}}</option>\r\n            <option value=\"18\">{{fontMiddle}}</option>\r\n            <option value=\"24\">{{fontBig}}</option>\r\n            <option value=\"36\">{{fontLarge}}</option>\r\n            <option value=\"48\">{{fontHuge}}</option>\r\n        </select>\r\n    </div>\r\n\r\n    <button *ngIf=\"cancelText && showCancelButton\" type=\"button\" class=\"btn btn-secondary  btn-lg\" (click)=\"cancelHandle()\">{{ cancelText }}</button>\r\n    <button #btnOkEl *ngIf=\"okText && showOkButton\" type=\"button\" class=\"btn btn-primary btn-lg\" (click)=\"okHandle()\">{{ okText }}</button>\r\n</div>\r\n    \r\n    \r\n<div class=\"modal-footer\" *ngIf=\"buttons && buttons.length\">\r\n    <span *ngIf=\"exception && exception.copyButton && exception.copyButton.text\" style=\"width: 100%;color: #2A87FF;padding-left: 37px;\">\r\n        <span (click)=\"onCopyClick($event)\" style=\"cursor: pointer;\">{{exception.copyButton.text}}</span></span>\r\n    <button type=\"button\"  *ngFor=\"let btn of buttons\"  (click)=\"btn.handle()\" [farris-default-focus]=\"btn\" [class]=\"btn.cls\"><span [class]=\"btn.iconCls\" *ngIf=\"btn.iconCls\"></span>{{ btn.text}}</button>\r\n</div>\r\n\r\n\r\n<div style=\"position: absolute;left: -21990px;top: -20000px; word-break: break-all;\r\ncolor: #747B8B;font-size: .875rem;\" *ngIf=\"exception\" #allexcep>\r\n{{detail}}: <span [innerHTML]=\"exception?.message | safe: 'html'\"></span>\r\n</div>\r\n\r\n<div style=\"position: absolute;\r\nleft: 50%;\r\ntop: 50%;\r\nwidth: 100px;\r\nheight: 40px;\r\nbackground: #303C53;\r\nline-height: 40px;\r\ntext-align: center;\r\nmargin-left: -30px;\r\nmargin-top: -50px;\r\nborder-radius: 10px;\r\nbox-shadow: 0px 2px 8px 0px;\r\nopacity: 0;\r\ncolor: #fff;\r\ntransition: all .3s ease; display: none\" #miniNotify>\r\n\r\n</div>\r\n\r\n"
                }] }
    ];
    /** @nocollapse */
    MessagerComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    MessagerComponent.propDecorators = {
        type: [{ type: Input }],
        showFontSize: [{ type: Input }],
        message: [{ type: Input }],
        msg: [{ type: Input }],
        cls: [{ type: HostBinding, args: ['class',] }],
        okText: [{ type: Input }],
        cancelText: [{ type: Input }],
        showOkButton: [{ type: Input }],
        showCancelButton: [{ type: Input }],
        readonly: [{ type: Input }],
        okHandle: [{ type: Input }],
        cancelHandle: [{ type: Input }],
        enableWordCount: [{ type: Input }],
        countType: [{ type: Input }],
        maxLength: [{ type: Input }],
        exception: [{ type: Input }],
        showLines: [{ type: Input }],
        promptText: [{ type: ViewChild, args: ['promptText',] }],
        fontSizeEl: [{ type: ViewChild, args: ['fontSizeEl',] }],
        wordCountEl: [{ type: ViewChild, args: ['wordCount',] }],
        excepMsgEl: [{ type: ViewChild, args: ['excepMsg',] }],
        allexcepEl: [{ type: ViewChild, args: ['allexcep',] }],
        miniNotifyEl: [{ type: ViewChild, args: ['miniNotify',] }],
        btnOkEl: [{ type: ViewChild, args: ['btnOkEl',] }],
        fontSize: [{ type: Input }],
        inputType: [{ type: Input }],
        placeholder: [{ type: Input }],
        buttons: [{ type: Input }]
    };
    return MessagerComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MessagerService = /** @class */ (function () {
    function MessagerService(bsModalService, injector) {
        this.bsModalService = bsModalService;
        this.injector = injector;
        this.version = "0.0.15";
        this.currentId = -1;
        this.modals = [];
        this.defaultMessagerConfig = null;
        /** @type {?} */
        var config = this.injector.get(MESSAGER_DEFAULT_CONFIG) || MessagerDefaultConfig;
        this.defaultMessagerConfig = Object.assign({}, MessagerDefaultConfig, config);
        this.ngZone = this.injector.get(NgZone);
        this.localeService = this.injector.get(LocaleService);
        this.commonUtils = this.injector.get(CommonUtils, null);
    }
    Object.defineProperty(MessagerService.prototype, "okText", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue("messager.ok");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerService.prototype, "cancelText", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue("messager.cancel");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerService.prototype, "yesText", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue("messager.yes");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerService.prototype, "noText", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue("messager.no");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerService.prototype, "titleText", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue("messager.title");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessagerService.prototype, "errorTitle", {
        get: /**
         * @return {?}
         */
        function () {
            return this.localeService.getValue("messager.errorTitle");
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} str
     * @return {?}
     */
    MessagerService.prototype.escapeHtml = /**
     * @param {?} str
     * @return {?}
     */
    function (str) {
        if (str === null || str === undefined) {
            return "";
        }
        return str
            .replace(/&/g, "&amp;")
            .replace(/</g, "&lt;")
            .replace(/>/g, "&gt;")
            .replace(/\"/g, "&quot;")
            .replace(/\'/g, "&#39;")
            .replace(/\//g, "&#x2F;");
    };
    /**
     * 显示信息提示框
     * @param msgtype 消息类别
     * @param message 提示消息一
     * @param opts 参数 提供属性 fitContent:true支持自动适应宽度
     * @param submsg 提示消息二级
     */
    /**
     * 显示信息提示框
     * @param {?} msgtype 消息类别
     * @param {?} message 提示消息一
     * @param {?=} opts 参数 提供属性 fitContent:true支持自动适应宽度
     * @param {?=} submsg 提示消息二级
     * @return {?}
     */
    MessagerService.prototype.show = /**
     * 显示信息提示框
     * @param {?} msgtype 消息类别
     * @param {?} message 提示消息一
     * @param {?=} opts 参数 提供属性 fitContent:true支持自动适应宽度
     * @param {?=} submsg 提示消息二级
     * @return {?}
     */
    function (msgtype, message, opts, submsg) {
        var _this = this;
        if (submsg === void 0) { submsg = ""; }
        if (document.activeElement) {
            ((/** @type {?} */ (document.activeElement))).blur();
        }
        /** @type {?} */
        var modalHeight = 180;
        /** @type {?} */
        var defaultOpts = {
            safeHtml: true,
            title: this.titleText,
            width: 390,
            height: modalHeight,
            showButtons: false,
            showMaxButton: false,
            resizable: false,
            initialState: {
                okText: this.okText,
                okHandle: (/**
                 * @return {?}
                 */
                function () {
                    _this.close();
                }),
            },
            class: "modal-message " +
                (msgtype ? "modal-message-type-" + msgtype : ""),
            showHeader: false,
        };
        defaultOpts = Object.assign({}, this.defaultMessagerConfig, defaultOpts);
        /** @type {?} */
        var extendOpts = {};
        switch (msgtype) {
            case "prompt":
                // 弹出对话类型
                extendOpts = { showHeader: true, modalHeight: 260 };
                break;
            case "error":
                // 错误
                extendOpts = {
                    title: this.errorTitle,
                    width: 446,
                    showHeader: true,
                    showMaxButton: true,
                    fitContent: true,
                };
                break;
            case "exception":
                extendOpts = {
                    safeHtml: false,
                    title: this.errorTitle,
                    width: 446,
                    showHeader: true,
                    fitContent: true,
                };
                break;
            default:
                // 默认其他
                extendOpts = { width: submsg ? 446 : 390, fitContent: true };
        }
        Object.assign(defaultOpts, extendOpts);
        if (opts) {
            Object.assign(defaultOpts, opts);
        }
        if (message && defaultOpts.safeHtml) {
            message = this.escapeHtml(message);
        }
        Object.assign(defaultOpts.initialState, {
            type: msgtype,
            message: message,
            msg: submsg,
            exception: opts && opts["exception"] ? opts["exception"] : null,
            showLines: opts && opts["showLines"] ? opts["showLines"] : 3,
            readonly: opts && opts["readonly"] ? opts["readonly"] : false,
        });
        if (defaultOpts["buttons"] && defaultOpts["buttons"].length) {
            if (!defaultOpts.initialState["buttons"] ||
                !defaultOpts.initialState["buttons"].length) {
                defaultOpts.initialState["buttons"] = defaultOpts["buttons"];
            }
        }
        /** @type {?} */
        var _modal = this.bsModalService.show(MessagerComponent, defaultOpts);
        // _modal.content.buttons = defaultOpts['buttons'] || [];
        this.confirmSubject = new Subject();
        if (this.modals.length > 0) {
            /** @type {?} */
            var modalContainerComponent = _modal.dialog;
            modalContainerComponent.instance.isShown = true;
            modalContainerComponent.instance.moveTo(15 * this.modals.length, 15 * this.modals.length);
        }
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            // 第1个按钮获取焦点
            setTimeout((/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var btnOk = _modal.content.btnOkEl
                    ? _modal.content.btnOkEl.nativeElement
                    : null;
                if (btnOk) {
                    btnOk["focus"]();
                }
                else {
                    /** @type {?} */
                    var btns = _modal.dialog.location.nativeElement.querySelectorAll(".modal-footer button");
                    if (btns.length) {
                        btns.item(0)["focus"]();
                    }
                }
                // const overlays = Array.from(document.body.querySelectorAll('body>div, body>farris-dialog>.farris-modal,body>.farris-modal')).filter(n => n).map(n => {
                //     return window.getComputedStyle(n).zIndex;
                // }).filter(n => n && n!== 'auto').map(n => {
                //     return parseInt(n, 10);
                // });
                /** @type {?} */
                var maxZindex = _this.commonUtils.getFloatingLayerIndex();
                _modal.dialog.location.nativeElement.style.zIndex = maxZindex;
            }));
        }));
        this.msgMgr(_modal);
        return _modal;
    };
    /**
     * @private
     * @param {?} ref
     * @return {?}
     */
    MessagerService.prototype.msgMgr = /**
     * @private
     * @param {?} ref
     * @return {?}
     */
    function (ref) {
        var _this = this;
        /** @type {?} */
        var id = this.bsModalService.getModalsCount() + 1000;
        if (this.modals.length === 0) {
            this.modals.push({ id: id, ref: ref });
        }
        else {
            if (!this.getModalById(id)) {
                this.modals.push({ id: id, ref: ref });
            }
            else {
                this.getModalById(id).ref = ref;
            }
        }
        /** @type {?} */
        var cmp = this.getModalCmp();
        cmp.closed.subscribe((/**
         * @param {?} level
         * @return {?}
         */
        function (level) {
            _this.removeMessager(id);
            _this.confirmSubject.unsubscribe();
        }));
        this.currentId = id;
    };
    /**
     * @param {?} message
     * @param {?} okCallback
     * @param {?=} cancelCallback
     * @param {?=} msg
     * @param {?=} fitContent
     * @return {?}
     */
    MessagerService.prototype.question = /**
     * @param {?} message
     * @param {?} okCallback
     * @param {?=} cancelCallback
     * @param {?=} msg
     * @param {?=} fitContent
     * @return {?}
     */
    function (message, okCallback, cancelCallback, msg, fitContent) {
        if (fitContent === void 0) { fitContent = true; }
        /** @type {?} */
        var questionDlg = this.show("question", message, {
            initialState: {
                okText: this.yesText,
                okHandle: (/**
                 * @return {?}
                 */
                function () {
                    questionDlg.close();
                    if (okCallback) {
                        okCallback();
                    }
                }),
                cancelText: this.noText,
                cancelHandle: (/**
                 * @return {?}
                 */
                function () {
                    questionDlg.close();
                    if (cancelCallback) {
                        cancelCallback();
                    }
                }),
            },
            fitContent: fitContent,
        }, msg);
        return questionDlg;
    };
    /**
     * @param {?} messager
     * @param {?} btns
     * @param {?=} msg
     * @param {?=} fitContent
     * @return {?}
     */
    MessagerService.prototype.question2 = /**
     * @param {?} messager
     * @param {?} btns
     * @param {?=} msg
     * @param {?=} fitContent
     * @return {?}
     */
    function (messager, btns, msg, fitContent) {
        if (fitContent === void 0) { fitContent = true; }
        return this.show("question", messager, {
            initialState: {
                buttons: btns,
            },
            fitContent: fitContent,
        }, msg);
    };
    /**
     * 2020年3月7日因为textarea高度在不同浏览器下不一样，固定高度可能会出现滚动条
     * @param title 标题
     * @param msg 信息
     * @param fitContent 自适应内容
     */
    /**
     * 2020年3月7日因为textarea高度在不同浏览器下不一样，固定高度可能会出现滚动条
     * @param {?} title 标题
     * @param {?=} msg 信息
     * @param {?=} fitContent 自适应内容
     * @return {?}
     */
    MessagerService.prototype.prompt = /**
     * 2020年3月7日因为textarea高度在不同浏览器下不一样，固定高度可能会出现滚动条
     * @param {?} title 标题
     * @param {?=} msg 信息
     * @param {?=} fitContent 自适应内容
     * @return {?}
     */
    function (title, msg, fitContent) {
        if (fitContent === void 0) { fitContent = true; }
        /** @type {?} */
        var promptSubject = new Subject();
        /** @type {?} */
        var dlg = this.show("prompt", msg, {
            title: title,
            initialState: {
                okText: this.okText,
                okHandle: (/**
                 * @return {?}
                 */
                function () {
                    /** @type {?} */
                    var mc = (/** @type {?} */ (dlg.content));
                    dlg.close();
                    promptSubject.next(mc.promptText.nativeElement.value);
                }),
                cancelText: this.cancelText,
                cancelHandle: (/**
                 * @return {?}
                 */
                function () {
                    dlg.close();
                    promptSubject.next(false);
                }),
            },
            fitContent: fitContent,
            closed: (/**
             * @return {?}
             */
            function () {
                promptSubject.unsubscribe();
            }),
        }, msg);
        return promptSubject;
    };
    /**
     * @param {?} title
     * @param {?=} msg
     * @param {?=} opts
     * @return {?}
     */
    MessagerService.prototype.prompt2 = /**
     * @param {?} title
     * @param {?=} msg
     * @param {?=} opts
     * @return {?}
     */
    function (title, msg, opts) {
        var _this = this;
        /** @type {?} */
        var promptSubject = new Subject();
        /** @type {?} */
        var _opts = {
            showOkButton: true,
            showCloseButton: true,
            showFontSize: false,
            fontSize: 18,
            readonly: false,
            saveSize: false,
            closeWhenever: true,
        };
        if (!opts) {
            opts = _opts;
        }
        else {
            opts = Object.assign(_opts, opts);
        }
        /** @type {?} */
        var _dialogOpts = {
            title: title,
            initialState: {
                showOkButton: opts.showOkButton === undefined ? true : opts.showOkButton,
                showCancelButton: opts.showCancelButton === undefined
                    ? true
                    : opts.showCancelButton,
                okText: this.okText,
                okHandle: (/**
                 * @return {?}
                 */
                function () {
                    /** @type {?} */
                    var mc = (/** @type {?} */ (dlg.content));
                    /** @type {?} */
                    var val = mc.promptText.nativeElement.value;
                    if (opts.closeWhenever) {
                        dlg.close();
                        promptSubject.next(val);
                    }
                    else {
                        promptSubject.next({ val: val, dlg: dlg });
                    }
                }),
                cancelText: this.cancelText,
                cancelHandle: (/**
                 * @return {?}
                 */
                function () {
                    dlg.close();
                    promptSubject.next(false);
                }),
                showFontSize: opts.showFontSize,
                fontSize: opts.fontSize || 18,
                enableWordCount: !!opts.enableWordCount,
                countType: opts.countType || "length",
                maxLength: opts.maxLength,
                inputType: opts.inputType || "textarea",
                placeholder: opts.placeholder || "",
            },
            fitContent: false,
            showMaxButton: true,
            resizable: true,
            class: "",
            closed: (/**
             * @return {?}
             */
            function () {
                promptSubject.unsubscribe();
            }),
            opened: (/**
             * @return {?}
             */
            function () {
                if (opts.inputType !== "textarea") {
                    return;
                }
                else {
                    setTimeout((/**
                     * @return {?}
                     */
                    function () {
                        /** @type {?} */
                        var textareaEle = dlg.content.promptText.nativeElement;
                        textareaEle.closest("section").style.padding = "15px";
                        textareaEle.style.resize = "none";
                        textareaEle.style.height = "100%";
                        if (opts.readonly) {
                            textareaEle.readOnly = true;
                        }
                    }));
                }
            }),
            beforeClose: (/**
             * @param {?} modalRef
             * @return {?}
             */
            function (modalRef) {
                if (opts.saveSize && !dlg.dialog.instance.isMax) {
                    /** @type {?} */
                    var data = {
                        fontSize: dlg.content.fontSize,
                        width: modalRef.config.width,
                        height: modalRef.config.height,
                    };
                    _this.saveTextAreaSizeInfo(data);
                }
                return of(true);
            }),
        };
        /** @type {?} */
        var dialogOpts = _dialogOpts;
        if (opts) {
            dialogOpts = Object.assign(dialogOpts, opts);
        }
        /** @type {?} */
        var dlg = this.show("prompt", msg, dialogOpts, msg);
        if (opts.inputType && dlg.content.promptText) {
            dlg.content.promptText.nativeElement.focus();
        }
        return promptSubject;
    };
    /**
     * @return {?}
     */
    MessagerService.prototype.getKeyString = /**
     * @return {?}
     */
    function () {
        // 一台终端共用一个配置
        // 保留最后一次的设置，
        // 保存窗口大小，字体大小
        return "IGIX-FARRIS-UI-TEXTAREA-ZOOM-SETTINGS";
    };
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    MessagerService.prototype.saveTextAreaSizeInfo = /**
     * @private
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (window.localStorage) {
            /** @type {?} */
            var key = this.getKeyString();
            localStorage.setItem(key, JSON.stringify(data));
        }
    };
    /**
     * @param {?} message
     * @param {?=} msg
     * @param {?=} fitContent
     * @return {?}
     */
    MessagerService.prototype.confirm = /**
     * @param {?} message
     * @param {?=} msg
     * @param {?=} fitContent
     * @return {?}
     */
    function (message, msg, fitContent) {
        if (fitContent === void 0) { fitContent = true; }
        /** @type {?} */
        var confirmSubject = new Subject();
        /** @type {?} */
        var _msg = typeof msg === "number" ? null : msg;
        /** @type {?} */
        var confirmDlg = this.show("question", message, {
            initialState: {
                okText: this.okText,
                okHandle: (/**
                 * @return {?}
                 */
                function () {
                    confirmDlg.close();
                    confirmSubject.next(true);
                }),
                cancelText: this.cancelText,
                cancelHandle: (/**
                 * @return {?}
                 */
                function () {
                    confirmDlg.close();
                    confirmSubject.next(false);
                }),
            },
            fitContent: fitContent,
            closed: (/**
             * @return {?}
             */
            function () {
                confirmSubject.unsubscribe();
            }),
        }, _msg);
        if (msg && typeof msg === "number") {
            confirmDlg.dialog.location.nativeElement.style.zIndex = msg;
        }
        return confirmSubject.asObservable();
    };
    /**
     * @param {?} message
     * @param {?=} callback
     * @param {?=} msg
     * @param {?=} fitContent
     * @return {?}
     */
    MessagerService.prototype.info = /**
     * @param {?} message
     * @param {?=} callback
     * @param {?=} msg
     * @param {?=} fitContent
     * @return {?}
     */
    function (message, callback, msg, fitContent) {
        if (fitContent === void 0) { fitContent = true; }
        /** @type {?} */
        var infoDlg = this.show("info", message, {
            initialState: {
                okText: this.okText,
                okHandle: (/**
                 * @return {?}
                 */
                function () {
                    infoDlg.close();
                    if (callback) {
                        callback();
                    }
                }),
            },
            fitContent: fitContent,
        }, msg);
        return infoDlg;
    };
    /** 操作成功提示，
     * message: 提示信息
     * msg: 2级信息
     * fitCountent: 默认为true
     * callback: 确定事件回调
     */
    /**
     * 操作成功提示，
     * message: 提示信息
     * msg: 2级信息
     * fitCountent: 默认为true
     * callback: 确定事件回调
     * @param {?} message
     * @param {?=} msg
     * @param {?=} fitContent
     * @param {?=} callback
     * @return {?}
     */
    MessagerService.prototype.success = /**
     * 操作成功提示，
     * message: 提示信息
     * msg: 2级信息
     * fitCountent: 默认为true
     * callback: 确定事件回调
     * @param {?} message
     * @param {?=} msg
     * @param {?=} fitContent
     * @param {?=} callback
     * @return {?}
     */
    function (message, msg, fitContent, callback) {
        if (fitContent === void 0) { fitContent = true; }
        if (callback === void 0) { callback = null; }
        /** @type {?} */
        var successDlg = this.show("success", message, {
            initialState: {
                okText: this.okText,
                okHandle: (/**
                 * @return {?}
                 */
                function () {
                    successDlg.close();
                    if (callback) {
                        callback();
                    }
                }),
            },
            fitContent: fitContent,
        }, msg);
        return successDlg;
    };
    /** 提示错误信息，
     * message: 提示信息
     * msg: 2级信息
     * fitCountent: 默认为true
     * callback: 确定事件回调
     */
    /**
     * 提示错误信息，
     * message: 提示信息
     * msg: 2级信息
     * fitCountent: 默认为true
     * callback: 确定事件回调
     * @param {?} message
     * @param {?=} msg
     * @param {?=} fitContent
     * @param {?=} callback
     * @param {?=} safeHtml
     * @return {?}
     */
    MessagerService.prototype.error = /**
     * 提示错误信息，
     * message: 提示信息
     * msg: 2级信息
     * fitCountent: 默认为true
     * callback: 确定事件回调
     * @param {?} message
     * @param {?=} msg
     * @param {?=} fitContent
     * @param {?=} callback
     * @param {?=} safeHtml
     * @return {?}
     */
    function (message, msg, fitContent, callback, safeHtml) {
        if (fitContent === void 0) { fitContent = true; }
        if (callback === void 0) { callback = null; }
        if (safeHtml === void 0) { safeHtml = true; }
        /** @type {?} */
        var errorDlg = this.show("error", message, {
            safeHtml: safeHtml,
            initialState: {
                okText: this.okText,
                okHandle: (/**
                 * @return {?}
                 */
                function () {
                    errorDlg.close();
                    if (callback) {
                        callback();
                    }
                }),
            },
            fitContent: fitContent,
        }, msg);
        return errorDlg;
    };
    /**
     * @param {?} message
     * @param {?=} msg
     * @param {?=} fitContent
     * @param {?=} callback
     * @return {?}
     */
    MessagerService.prototype.warning = /**
     * @param {?} message
     * @param {?=} msg
     * @param {?=} fitContent
     * @param {?=} callback
     * @return {?}
     */
    function (message, msg, fitContent, callback) {
        if (fitContent === void 0) { fitContent = true; }
        if (callback === void 0) { callback = null; }
        /** @type {?} */
        var warningDlg = this.show("warning", message, {
            initialState: {
                okText: this.okText,
                okHandle: (/**
                 * @return {?}
                 */
                function () {
                    warningDlg.close();
                    if (callback) {
                        callback();
                    }
                }),
            },
            fitContent: fitContent,
        }, msg);
        return warningDlg;
    };
    /**
     * @param {?} type
     * @param {?} title
     * @param {?=} message
     * @param {?=} callback
     * @return {?}
     */
    MessagerService.prototype.showHtmlMsg = /**
     * @param {?} type
     * @param {?} title
     * @param {?=} message
     * @param {?=} callback
     * @return {?}
     */
    function (type, title, message, callback) {
        if (callback === void 0) { callback = null; }
        /** @type {?} */
        var msgDlg = this.show(type, title, {
            safeHtml: false,
            initialState: {
                okText: this.okText,
                okHandle: (/**
                 * @return {?}
                 */
                function () {
                    msgDlg.close();
                    if (callback) {
                        callback();
                    }
                }),
            },
            fitContent: true,
        }, message);
        return msgDlg;
    };
    /**
     * @private
     * @return {?}
     */
    MessagerService.prototype.close = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.modals.length) {
            /** @type {?} */
            var msgbox = this.getModalById(this.currentId);
            if (msgbox) {
                msgbox.ref.close();
            }
        }
    };
    /**
     * @private
     * @param {?} id
     * @return {?}
     */
    MessagerService.prototype.getModalById = /**
     * @private
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.modals.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.id === id; }));
    };
    /**
     * @private
     * @return {?}
     */
    MessagerService.prototype.getModalCmp = /**
     * @private
     * @return {?}
     */
    function () {
        return this.bsModalService.getCurrentModalContainer().instance;
    };
    /**
     * @private
     * @param {?} id
     * @return {?}
     */
    MessagerService.prototype.removeMessager = /**
     * @private
     * @param {?} id
     * @return {?}
     */
    function (id) {
        this.modals = this.modals.filter((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n.id !== id; }));
        if (!this.modals.length) {
            this.currentId = -1;
        }
        else {
            this.currentId = this.modals[this.modals.length - 1].id;
        }
    };
    /**
     * @private
     * @param {?} text
     * @return {?}
     */
    MessagerService.prototype.copyToClipboard = /**
     * @private
     * @param {?} text
     * @return {?}
     */
    function (text) {
        /** @type {?} */
        var windowRef = window;
        if (windowRef.clipboardData && windowRef.clipboardData.setData) {
            return windowRef.clipboardData.setData("Text", text);
        }
        else if (document.queryCommandSupported && document.queryCommandSupported("copy")) {
            /** @type {?} */
            var textarea = document.createElement("textarea");
            textarea.textContent = text;
            textarea.style.position = "fixed";
            document.body.appendChild(textarea);
            textarea.select();
            try {
                return document.execCommand("copy");
            }
            catch (ex) {
            }
            finally {
                document.body.removeChild(textarea);
            }
        }
        return false;
    };
    /** 将请求异常信息体格式化为object */
    /**
     * 将请求异常信息体格式化为object
     * @private
     * @param {?} requestFailedInfo
     * @return {?}
     */
    MessagerService.prototype.getServerExceptionEntity = /**
     * 将请求异常信息体格式化为object
     * @private
     * @param {?} requestFailedInfo
     * @return {?}
     */
    function (requestFailedInfo) {
        if (requestFailedInfo.status == 404) {
            /** @type {?} */
            var serverExceptionEntity_1 = (/** @type {?} */ ({}));
            serverExceptionEntity_1.Message = (requestFailedInfo && requestFailedInfo.message) || 'Request Failed!';
            return serverExceptionEntity_1;
        }
        /** @type {?} */
        var serverExceptionEntity = requestFailedInfo && requestFailedInfo.error;
        if (serverExceptionEntity) {
            switch (typeof serverExceptionEntity) {
                case 'string':
                    return JSON.parse(serverExceptionEntity);
                case 'object':
                    return serverExceptionEntity;
                default:
                    console.error('不支持的请求异常数据格式：' + typeof serverExceptionEntity);
                    console.error(serverExceptionEntity);
                    return null;
            }
        }
        else {
            return null;
        }
    };
    /**
     * @param {?} _error
     * @param {?=} opts
     * @return {?}
     */
    MessagerService.prototype.exceptionWithHttpInfo = /**
     * @param {?} _error
     * @param {?=} opts
     * @return {?}
     */
    function (_error, opts) {
        if (opts === void 0) { opts = { width: 440, height: 200, log: false }; }
        /** @type {?} */
        var error = this.getServerExceptionEntity(_error);
        this.exception(error, opts);
    };
    /**
     * @param {?} error
     * @param {?=} opts
     * @return {?}
     */
    MessagerService.prototype.exception = /**
     * @param {?} error
     * @param {?=} opts
     * @return {?}
     */
    function (error, opts) {
        var _this = this;
        if (opts === void 0) { opts = { width: 440, height: 200, log: false }; }
        /** @type {?} */
        var exceptionMsg = error.Message || "";
        /** @type {?} */
        var date = error.date || format(new Date(), 'yyyy-MM-dd HH:mm:ss');
        /** @type {?} */
        var localDate = date;
        // 发生时间
        /** @type {?} */
        var happenedTime = this.localeService.getValue('messager.exception.happend');
        // 详细信息
        /** @type {?} */
        var detailInfo = this.localeService.getValue('messager.exception.detail');
        /** @type {?} */
        var detail = error.Detail ||
            error.Message + "\r\n" + happenedTime + "\uFF1A" + localDate + "\r\n" + detailInfo + "\uFF1A" + error.innerMessage ||
            "";
        /** @type {?} */
        var innerMessage = error.innerMessage || null;
        /** @type {?} */
        var defaultOptions = { width: 440, height: 200, log: false };
        /** @type {?} */
        var exoptions = defaultOptions;
        if (opts) {
            exoptions = Object.assign({}, defaultOptions, opts);
        }
        /** @type {?} */
        var bsModalRef;
        /** @type {?} */
        var options = {
            showMaxButton: false,
            buttons: [
                {
                    text: this.localeService.getValue('messager.exception.roger'),
                    cls: "btn btn-primary btn-lg",
                    handle: (/**
                     * @return {?}
                     */
                    function () {
                        bsModalRef.close();
                    }),
                },
            ],
            // showHeader: false,
            width: exoptions.width,
            height: exoptions.height,
            safeHtml: false,
            exception: {
                date: localDate,
                message: innerMessage,
                copyButton: {
                    text: this.localeService.getValue('messager.exception.copy'),
                    onClick: (/**
                     * @param {?} e
                     * @return {?}
                     */
                    function (e) {
                        /** @type {?} */
                        var result = _this.copyToClipboard(detail);
                        /** @type {?} */
                        var text = result
                            ? _this.localeService.getValue('messager.exception.copySuccess')
                            : _this.localeService.getValue('messager.exception.copyFailed');
                        bsModalRef.content.showMiniNotify(text, 1500);
                    }),
                },
            },
        };
        if (exoptions.log) {
            console.error(error);
        }
        // this.displayError(error);
        bsModalRef = this.show("exception", exceptionMsg, options);
    };
    MessagerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MessagerService.ctorParameters = function () { return [
        { type: BsModalService },
        { type: Injector }
    ]; };
    return MessagerService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MessagerModule = /** @class */ (function () {
    function MessagerModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    MessagerModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: MessagerModule,
            providers: [
                { provide: MESSAGER_DEFAULT_CONFIG, useValue: config || MessagerDefaultConfig },
                MessagerService
            ]
        };
    };
    MessagerModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        FarrisCommonModule.forRoot(),
                        // FarrisFormsModule,
                        LocaleModule,
                        ModalModule.forRoot()
                    ],
                    declarations: [
                        FarrisDefaultFocusDirective,
                        MessagerComponent
                    ],
                    entryComponents: [
                        MessagerComponent
                    ]
                },] }
    ];
    return MessagerModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { FarrisDefaultFocusDirective, MessagerDefaultConfig, MESSAGER_DEFAULT_CONFIG, MessagerService, MessagerComponent, MessagerModule };
