// @gsp-sys/sysmgr-ui/esm5/lib/sys-manager-ui/service/sys-help.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { from } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { SysCommonUtilService, SysDataUtil } from '@gsp-sys/sysmgr-common';
import { map } from 'rxjs/operators';
var SysHelpService = /** @class */ (function () {
    function SysHelpService(http, sysUtil) {
        this.http = http;
        this.sysUtil = sysUtil;
    }
    /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    SysHelpService.prototype.getHttpOption = /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        SysDataUtil.deleteEmpty(filter);
        /** @type {?} */
        var paramStr = JSON.stringify(filter);
        /** @type {?} */
        var opt = {
            headers: this.sysUtil.getHeader()['headers'],
            params: { param: paramStr }
        };
        return opt;
    };
    /**
     * @param {?} filter
     * @return {?}
     */
    SysHelpService.prototype.getUsers = /**
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        /** @type {?} */
        var opt = this.getHttpOption(filter);
        // 用户目前的有分页信息
        return this.http.get(this.sysUtil.userWebApi, opt).pipe(map((/**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            return val.data;
        })));
    };
    /**
     * @param {?} filter
     * @return {?}
     */
    SysHelpService.prototype.getPositions = /**
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        /** @type {?} */
        var opt = this.getHttpOption(filter);
        return this.http.get(this.sysUtil.positionWebApi, opt).pipe(map((/**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            return val.data;
        })));
    };
    /**
     * @param {?} filter
     * @return {?}
     */
    SysHelpService.prototype.getRoles = /**
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        /** @type {?} */
        var opt = this.getHttpOption(filter);
        return this.http.get(this.sysUtil.roleWebApi, opt).pipe(map((/**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            return val.data;
        })));
    };
    /**
     * @param {?} filter
     * @return {?}
     */
    SysHelpService.prototype.getOrgs = /**
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        /** @type {?} */
        var opt = this.getHttpOption(filter);
        // 视图
        opt.params.useView = true;
        return this.handerTreeData(this.http.get(this.sysUtil.sysOrgWebApi + '/all', opt));
    };
    /**
     * @param {?} filter
     * @return {?}
     */
    SysHelpService.prototype.getUserGroups = /**
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        /** @type {?} */
        var opt = this.getHttpOption(filter);
        return this.handerTreeData(this.http.get(this.sysUtil.userGroupWebApi + '/all', opt));
    };
    /**
     * @param {?} filter
     * @return {?}
     */
    SysHelpService.prototype.getPositionGroups = /**
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        /** @type {?} */
        var opt = this.getHttpOption(filter);
        return this.handerTreeData(this.http.get(this.sysUtil.posGroupWebApi + '/all', opt));
    };
    /**
     * @param {?} filter
     * @return {?}
     */
    SysHelpService.prototype.getRoleGroups = /**
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        /** @type {?} */
        var opt = this.getHttpOption(filter);
        return this.handerTreeData(this.http.get(this.sysUtil.roleGroupWebApi + '/all', opt));
    };
    /**
     * @param {?} item
     * @return {?}
     */
    SysHelpService.prototype.convert2Json = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        /** @type {?} */
        var result = {
            id: item.id,
            code: item.code,
            name: item.name,
            parentId: item.parentId || '',
            path: item.path,
            layer: item.layer,
            isDetail: item.isDetail,
        };
        if (item.nodeType != null && item.nodeType !== undefined) {
            result.nodeType = item.nodeType;
        }
        return result;
    };
    /**
     * @param {?} org
     * @return {?}
     */
    SysHelpService.prototype.getParentOrg = /**
     * @param {?} org
     * @return {?}
     */
    function (org) {
        /** @type {?} */
        var paramStr = JSON.stringify(this.convert2Json(org));
        /** @type {?} */
        var opt = {
            headers: this.sysUtil.getHeader()['headers'],
            params: { param: paramStr }
        };
        return this.http.get(this.sysUtil.sysOrgWebApi + '/getparentorg/0002', opt).pipe(map((/**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            return val;
        })));
    };
    /** 适配Kendo-treeView结构 */
    /**
     * 适配Kendo-treeView结构
     * @private
     * @param {?} src
     * @return {?}
     */
    SysHelpService.prototype.handerTreeData = /**
     * 适配Kendo-treeView结构
     * @private
     * @param {?} src
     * @return {?}
     */
    function (src) {
        return src.pipe(map((/**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            from(val).subscribe((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.layer == 1 || item.parentId == null || item.parentId === '') {
                    item.parentId = undefined;
                }
            }));
            return val;
        })));
    };
    SysHelpService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SysHelpService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: SysCommonUtilService }
    ]; };
    return SysHelpService;
}());
export { SysHelpService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    SysHelpService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    SysHelpService.prototype.sysUtil;
}
