// @gsp-wf/wf-process-editor/esm5/lib/views/release-component/release-component.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, TemplateRef, ChangeDetectorRef, Optional, Inject, LOCALE_ID } from '@angular/core';
import { LookupGridComponent } from '@farris/ui-lookup';
import { WfTaskHandlerService } from '@gsp-wf/wf-task-handler';
import { ProcessDesignerUIState } from '../../app/process-designer.uistate';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../i18n/pipe/wf-process-editor-locale.pipe';
var ReleaseComponent = /** @class */ (function () {
    function ReleaseComponent(uiState, submitService, cdr, localeId, resources, localePipe) {
        this.uiState = uiState;
        this.submitService = submitService;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.submit = new EventEmitter();
        this.cancel = new EventEmitter();
        this.accessLevel = 'public';
        this.isAdministratorConfig = false;
        this.dataUri = '/api/runtime/sys/v1.0/sysOrgs';
        this.loadType = 'layerload';
        this.releaseType = 'release';
        this.items = new Array();
        this.releaseTypeList = [
            { value: 'release', name: this.localePipe.transform('component.release.release') },
            { value: 'update', name: this.localePipe.transform('component.release.update') },
        ];
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
    }
    /**
     * @return {?}
     */
    ReleaseComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.releaseType = this.procVersion === 1 ? 'release' : 'update';
        this.uiState.getUserInfo().subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            _this.applicantName = re.name;
            _this.applicantDepartment = re.orgName;
            _this.applicantId = re.id;
        }));
    };
    /**
     * @return {?}
     */
    ReleaseComponent.prototype.saveAsOK = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.uiState.createData().subscribe((/**
         * @param {?} saveData
         * @return {?}
         */
        function (saveData) {
            /** @type {?} */
            var dataId = saveData.returnValue.id;
            /** @type {?} */
            var data = {
                DataId: dataId,
                applicantName: _this.applicantName,
                applicantDepartment: _this.applicantDepartment,
                procName: _this.procName,
                procVersion: _this.procVersion,
                releaseType: _this.releaseType,
                textDetail: _this.textDetail,
                processDefinitionKey: _this.processDefinitionKey,
                applicantId: _this.applicantId
            };
            /** @type {?} */
            var Info = {
                ChangeType: 'Modify',
                ChangeInfo: data
            };
            _this.uiState.releaseSave(Info).subscribe((/**
             * @param {?} re
             * @return {?}
             */
            function (re) {
                _this.mySubmit(dataId, '03bdca37-8439-4924-8aef-20f3f4092b59');
                _this.submit.next();
            }));
        }));
    };
    /**
     * @return {?}
     */
    ReleaseComponent.prototype.saveAsCancel = /**
     * @return {?}
     */
    function () {
        this.cancel.emit();
    };
    /**
     * @param {?} quill
     * @return {?}
     */
    ReleaseComponent.prototype.editorCreated = /**
     * @param {?} quill
     * @return {?}
     */
    function (quill) {
        /** @type {?} */
        var toolbar = quill.getModule('toolbar');
        this.editor = quill;
    };
    /**
     * @param {?} v
     * @return {?}
     */
    ReleaseComponent.prototype.contentChanged = /**
     * @param {?} v
     * @return {?}
     */
    function (v) {
        console.log(v);
    };
    /**
     * @return {?}
     */
    ReleaseComponent.prototype.validator = /**
     * @return {?}
     */
    function () {
        return true;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    ReleaseComponent.prototype.changeValue = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.releaseType = data;
    };
    /**
     * 发布前审核：提交审批
     */
    /**
     * 发布前审核：提交审批
     * @private
     * @param {?} dataId
     * @param {?} bizDefKey
     * @return {?}
     */
    ReleaseComponent.prototype.mySubmit = /**
     * 发布前审核：提交审批
     * @private
     * @param {?} dataId
     * @param {?} bizDefKey
     * @return {?}
     */
    function (dataId, bizDefKey) {
        this.submitService.submitWithBizDefKey(dataId, bizDefKey);
    };
    ReleaseComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-release-component',
                    template: "<div class=\"f-page f-page-card f-page-is-mainsubcard\">\r\n  <section class=\"f-page-main\">\r\n    <div class=\"f-struct-wrapper\">\r\n      <!-- <farris-section class=\"f-section-form f-section-in-mainsubcard\" [enableAccordion]=\"''\"\r\n        [subTitle]=\"''\" [enableMaximize]=\"false\"> -->\r\n        <div class=\"f-form-layout farris-form farris-form-controls-inline pr-4 m-2\">\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{'component.release.applicantName' | wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <input type=\"text\" class=\"form-control\" [(ngModel)]=\"applicantName\" name=\"procName\" [readOnly]=\"true\"/>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{'component.release.applicantDepartment' | wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <input type=\"text\" class=\"form-control\" [(ngModel)]=\"applicantDepartment\" name=\"procKey\" [readOnly]=\"true\"/>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{'component.release.procName' | wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <input type=\"text\" class=\"form-control\" [(ngModel)]=\"procName\" name=\"procName\" [readOnly]=\"true\"/>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{'component.release.procVersion' | wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <input type=\"text\" class=\"form-control\" [(ngModel)]=\"procVersion\" name=\"procVersion\" [readOnly]=\"true\"/>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{'component.release.releaseType' | wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <farris-radiogroup [(ngModel)]=\"releaseType\" name=\"releaseType\" [data]=\"releaseTypeList\"\r\n                  [horizontal]=\"true\" (changeValue)=\"changeValue($event)\" [disabled]=\"true\">\r\n                  </farris-radiogroup>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{'component.release.textDetail' | wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <textarea [(ngModel)]=\"textDetail\" name=\"textarea\" id=\"textarea\" style=\"height: 250px; width: 100%\"\r\n                  [auto-size]=\"true\" [maxHeight]=\"500\" [textarea-zoom]=\"true\"></textarea>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      <!-- </farris-section> -->\r\n    </div>\r\n  </section>\r\n</div>\r\n\r\n<ng-template #buttonRef>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm mx-2\" (click)=\"saveAsCancel()\">{{'component.release.saveAsCancel' | wfLocale}}</button>\r\n  <button type=\"button\" class=\"btn btn-sm btn-primary mx-2\" (click)=\"saveAsOK()\">{{'component.release.saveAsOK' | wfLocale}}</button>\r\n</ng-template>\r\n",
                    providers: [
                        WfLocalePipe,
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ReleaseComponent.ctorParameters = function () { return [
        { type: ProcessDesignerUIState },
        { type: WfTaskHandlerService },
        { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
        { type: WfLocalePipe, decorators: [{ type: Optional }] }
    ]; };
    ReleaseComponent.propDecorators = {
        btnRef: [{ type: ViewChild, args: ['buttonRef',] }],
        procName: [{ type: Input }],
        procVersion: [{ type: Input }],
        processDefinitionKey: [{ type: Input }],
        lookUp: [{ type: ViewChild, args: ['lookup',] }],
        submit: [{ type: Output }],
        cancel: [{ type: Output }]
    };
    return ReleaseComponent;
}());
export { ReleaseComponent };
if (false) {
    /** @type {?} */
    ReleaseComponent.prototype.btnRef;
    /** @type {?} */
    ReleaseComponent.prototype.applicantName;
    /** @type {?} */
    ReleaseComponent.prototype.applicantDepartment;
    /** @type {?} */
    ReleaseComponent.prototype.procName;
    /** @type {?} */
    ReleaseComponent.prototype.procVersion;
    /** @type {?} */
    ReleaseComponent.prototype.processDefinitionKey;
    /** @type {?} */
    ReleaseComponent.prototype.lookUp;
    /** @type {?} */
    ReleaseComponent.prototype.submit;
    /** @type {?} */
    ReleaseComponent.prototype.cancel;
    /** @type {?} */
    ReleaseComponent.prototype.session;
    /** @type {?} */
    ReleaseComponent.prototype.accessLevel;
    /** @type {?} */
    ReleaseComponent.prototype.organization;
    /** @type {?} */
    ReleaseComponent.prototype.organizationName;
    /** @type {?} */
    ReleaseComponent.prototype.textDetail;
    /** @type {?} */
    ReleaseComponent.prototype.isAdministratorConfig;
    /** @type {?} */
    ReleaseComponent.prototype.editor;
    /** @type {?} */
    ReleaseComponent.prototype.administrator;
    /** @type {?} */
    ReleaseComponent.prototype.dataUri;
    /** @type {?} */
    ReleaseComponent.prototype.applicantId;
    /** @type {?} */
    ReleaseComponent.prototype.loadType;
    /** @type {?} */
    ReleaseComponent.prototype.releaseType;
    /** @type {?} */
    ReleaseComponent.prototype.items;
    /** @type {?} */
    ReleaseComponent.prototype.releaseTypeList;
    /**
     * @type {?}
     * @private
     */
    ReleaseComponent.prototype.uiState;
    /**
     * @type {?}
     * @private
     */
    ReleaseComponent.prototype.submitService;
    /**
     * @type {?}
     * @private
     */
    ReleaseComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    ReleaseComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    ReleaseComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    ReleaseComponent.prototype.localePipe;
}
