// @farris/command-services/esm5/lib/query-solution/query-solution.service.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { QueryConditionHandlerFactory } from './querycondition.handler.factory';
import { FrameContext } from '@farris/devkit';
import { FrameworkSessionService } from '@farris/bef';
var QuerySolutionService = /** @class */ (function () {
    function QuerySolutionService(frameContext, frameworkSessionService) {
        this.frameContext = frameContext;
        this.frameworkSessionService = frameworkSessionService;
    }
    QuerySolutionService.prototype.getUserSessionId = function () {
        return this.frameworkSessionService.getUserSessionId();
    };
    QuerySolutionService.prototype.setFilterConditions = function (params) {
        var queryConditionHandler;
        var filterConditionList = [];
        params.forEach(function (qc) {
            if (!qc.value.isEmpty()) {
                queryConditionHandler = QueryConditionHandlerFactory.getInstance().getHandler(qc.control.getControlType());
                queryConditionHandler && filterConditionList.push.apply(filterConditionList, tslib_1.__spread(queryConditionHandler.convert(qc)));
            }
        });
        //设置到NgCommand params.filter属性上：{UISTATE~/root-component/filterConditionList}
        this.frameContext.uiState['filterConditionList'] = JSON.stringify(filterConditionList);
        // const originalConditions = this.getOriginalQueryConditions(filterConditionList);
        this.frameContext.uiState['originalFilterConditionList'] = JSON.stringify(filterConditionList);
    };
    QuerySolutionService.prototype.setCurrentQueryConditions = function (conditions) {
        this.frameContext.uiState['currentQueryConditions'] = conditions;
    };
    QuerySolutionService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    QuerySolutionService.ctorParameters = function () { return [
        { type: FrameContext },
        { type: FrameworkSessionService }
    ]; };
    return QuerySolutionService;
}());
export { QuerySolutionService };
