// @gsp-lcm/bo-creator/esm5/lib/views/bo-creator/modal/bo-modal/bo-creator-modal.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/views/bo-creator/modal/bo-modal/bo-creator-modal.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output, ViewChild } from '@angular/core';
import { ComboListComponent } from '@farris/ui-combo-list';
import { BusinessObjectEntity, BusinessObjectDtService } from '@gsp-lcm/bo-dt-service';
import { BoCreatorUtils } from '../../utils/bo-create-utils';
import { NotifyService } from '@farris/ui-notify';
import { ValidatorResult } from '../../common/validation-result';
var BoCreatorModalComponent = /** @class */ (function () {
    function BoCreatorModalComponent(boCreatorUtils, boDtService, notifyService) {
        this.boCreatorUtils = boCreatorUtils;
        this.boDtService = boDtService;
        this.notifyService = notifyService;
        this.boConfirmEmitter = new EventEmitter();
        /**
         * 当前BO节点
         */
        this.currentBo = new BusinessObjectEntity();
        /**
         * 关键应用集合
         */
        this.items = [];
        /**
         * 选择应用节点下的所有服务单元集合
         */
        this.suList = [];
        /**
         * 所有业务对象节点集合
         */
        this.data = [];
        this.parentBo = new BusinessObjectEntity();
        this.labelLength = '2';
        this.inputLength = '9';
    }
    Object.defineProperty(BoCreatorModalComponent.prototype, "_parentId", {
        set: /**
         * @param {?} parentId
         * @return {?}
         */
        function (parentId) {
            this.parentId = parentId;
            this.getParentBo();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initData();
    };
    /**
     * @private
     * @return {?}
     */
    BoCreatorModalComponent.prototype.initData = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.boDtService.getBusinessObjects().subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.data = data;
            _this.items = data.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.parentID === _this.parentId && item.layer === 4 && item.isDetail === '0'; }));
            /** @type {?} */
            var children = data.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.parentID === _this.parentId; }));
            /** @type {?} */
            var boInSU = [];
            children.forEach((/**
             * @param {?} s
             * @return {?}
             */
            function (s) {
                if (s.isDetail == '0') {
                    boInSU.push(s);
                    /** @type {?} */
                    var bos = data.filter((/**
                     * @param {?} t
                     * @return {?}
                     */
                    function (t) { return t.parentID === s.id; }));
                    if (bos) {
                        bos.forEach((/**
                         * @param {?} bo
                         * @return {?}
                         */
                        function (bo) {
                            boInSU.push(bo);
                        }));
                    }
                }
                else {
                    boInSU.push(s);
                }
            }));
            _this.bosInSU = boInSU;
        }));
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.create = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var bo = this.getCurrentBo();
        if (!bo) {
            return;
        }
        this.boDtService.save(bo).subscribe((/**
         * @return {?}
         */
        function () {
            _this.boConfirmEmitter.emit(bo);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.notifyService.error('创建失败');
            throw error;
        }));
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.getCurrentBo = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = this.boInfoValidator();
        if (result.illegal) {
            this.notifyService.warning(result.message);
            return null;
        }
        this.currentBo.id = this.boCreatorUtils.newGuid();
        this.currentBo.layer = this.parentBo ? this.parentBo.layer + 1 : 4;
        this.currentBo.parentID = this.parentBo && this.parentBo.id ? this.parentBo.id : this.parentId;
        this.currentBo.sysInit = '0';
        this.currentBo.isSysInit = false;
        this.currentBo.isDetail = '1';
        /** @type {?} */
        var max = this.boCreatorUtils.getMaxOrder(this.bosInSU, this.currentBo.parentID);
        this.currentBo.sortOrder = max + 1;
        return Object.assign(new BusinessObjectEntity(), this.currentBo);
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.boInfoValidator = /**
     * @return {?}
     */
    function () {
        //新增信息空校验
        /** @type {?} */
        var result = this.emptyValidator();
        if (result.illegal) {
            return result;
        }
        //编号，名称命名规范性校验
        result = this.namingConventionValidator();
        if (result.illegal) {
            return result;
        }
        //APP编号重复校验
        return this.repeatValidator();
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.emptyValidator = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = new ValidatorResult();
        if (!this.suCode) {
            result.illegal = true;
            result.message = "\u6240\u5C5E\u5FAE\u670D\u52A1\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        if (!this.currentBo.code) {
            result.illegal = true;
            result.message = "\u4E1A\u52A1\u5BF9\u8C61\u7684\u7F16\u53F7\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        if (!this.currentBo.name) {
            result.illegal = true;
            result.message = "\u4E1A\u52A1\u5BF9\u8C61\u7684\u540D\u79F0\u4E0D\u80FD\u4E3A\u7A7A";
            return result;
        }
        return result;
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.namingConventionValidator = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result = new ValidatorResult();
        /** @type {?} */
        var codeRegx = new RegExp('^[a-zA-Z\x7f-\xff][a-zA-Z0-9\x7f-\xff]*$');
        if (!codeRegx.test(this.currentBo.code)) {
            result.illegal = true;
            result.message = '业务对象的编号只能包含字母,数字,并且不可以用数字开头!';
            return result;
        }
        /** @type {?} */
        var nameRegx = new RegExp('[`~!@#$^&*()=|{}\':;\',\\[\\].<>/?~！@#￥……&*（）——|{}【】‘；：”“\'。，、？]');
        if (nameRegx.test(this.currentBo.name)) {
            result.illegal = true;
            result.message = '业务对象的名称包含非法字符，请重新输入!';
            return result;
        }
        return result;
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.repeatValidator = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var result = new ValidatorResult();
        if (!this.bosInSU || this.bosInSU.length <= 0) {
            return result;
        }
        /** @type {?} */
        var repeatCodeBO = this.bosInSU.find((/**
         * @param {?} bo
         * @return {?}
         */
        function (bo) { return bo.code.toLowerCase() === _this.currentBo.code.toLowerCase(); }));
        if (repeatCodeBO) {
            result.illegal = true;
            result.message = '同一SU下业务对象/业务对象分组/业务对象与业务对象分组编号不能重复，业务对象编号为：' + this.currentBo.code;
            return result;
        }
        return result;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    BoCreatorModalComponent.prototype.boGroupSelectChanged = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.parentBo = $event.data;
    };
    /**
     * @private
     * @return {?}
     */
    BoCreatorModalComponent.prototype.getParentBo = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.parentId) {
            this.boDtService.getBusinessObject(this.parentId).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                _this.parentBo = data;
                _this.suCode = data.code;
            }));
        }
    };
    /**
     * @return {?}
     */
    BoCreatorModalComponent.prototype.clearHandler = /**
     * @return {?}
     */
    function () {
        this.parentBo = null;
    };
    BoCreatorModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-bo-creator-modal',
                    template: "<div class=\"cloumn\" style=\"margin-top: 28px;\">\n\n  <div class=\"suInfo row\">\n    <label class=\"col-form-label ps-col-form-label\">\n      <span class=\"farris-label-info text-danger\">*</span>\n      <span class=\"farris-label-text\">\u6240\u5C5E\u5FAE\u670D\u52A1</span>\n    </label>\n    <div class=\"col-{{inputLength}}\">\n      <input-group [(ngModel)]=\"suCode\" #boCode=\"ngModel\" [readonly]=\"true\"></input-group>\n    </div>\n  </div>\n\n  <div class=\"row suInfo\">\n    <label class=\"col-form-label ps-col-form-label\">\n      <span class=\"farris-label-text\">\u4E1A\u52A1\u5BF9\u8C61\u5206\u7EC4</span>\n    </label>\n    <div class=\"col-{{inputLength}}\">\n      <farris-combo-list #boGroupSelector [idField]=\"'id'\" [textField]=\"'name'\" [data]=\"items\"\n        (selectChange)=\"boGroupSelectChanged($event)\" (clear)=\"clearHandler()\" [panelHeight]=\"280\" [editable]=\"false\"></farris-combo-list>\n    </div>\n  </div>\n\n  <div class=\"row suInfo\">\n    <label class=\"col-form-label ps-col-form-label\">\n      <span class=\"farris-label-info text-danger\">*</span>\n      <span class=\"farris-label-text\">\u7F16\u53F7</span>\n    </label>\n    <div class=\"col-{{inputLength}}\">\n      <input-group [(ngModel)]=\"currentBo.code\" #boCode=\"ngModel\" required></input-group>\n    </div>\n    <!-- <div [hidden]=\"boCode.valid || boCode.pristine\" class=\"alert-danger\">\n      \u7F16\u53F7\u4E0D\u80FD\u4E3A\u7A7A\n    </div> -->\n  </div>\n\n  <div class=\"row suInfo\">\n    <label class=\"col-form-label ps-col-form-label\">\n      <span class=\"farris-label-info text-danger\">*</span>\n      <span class=\"farris-label-text\">\u540D\u79F0</span>\n    </label>\n    <div class=\"col-{{inputLength}}\">\n      <input-group [(ngModel)]=\"currentBo.name\" #boName=\"ngModel\" required></input-group>\n    </div>\n    <!-- <div [hidden]=\"boName.valid || boName.pristine\" class=\"alert-danger\">\n      \u540D\u79F0\u4E0D\u80FD\u4E3A\u7A7A\n    </div> -->\n  </div>\n\n</div>",
                    styles: [".sucreator-wrapper{background:rgba(255,255,255,.8);width:95%;margin:40px auto 0;max-width:480px;padding:24px 18px;border-radius:8px;position:relative;z-index:100}.suInfo{margin:0 0 16px 20px}.sucreator-wrapper .col-form-label,.sucreator-wrapper .form-control{padding-top:3px;padding-bottom:3px}.sucreator-wrapper .col-form-label{width:70px;margin-right:12px;text-align:right;flex-shrink:0}.bocreator{background:#fff;box-shadow:0 2px 16px 0 rgba(0,0,0,.1);border-radius:4px}.boCreatorPanel{background-color:#fff}div .ps-col-form-label{width:94px;text-align:right}"]
                }] }
    ];
    /** @nocollapse */
    BoCreatorModalComponent.ctorParameters = function () { return [
        { type: BoCreatorUtils },
        { type: BusinessObjectDtService },
        { type: NotifyService }
    ]; };
    BoCreatorModalComponent.propDecorators = {
        _parentId: [{ type: Input }],
        boConfirmEmitter: [{ type: Output }],
        boGroupSelector: [{ type: ViewChild, args: ["boGroupSelector",] }]
    };
    return BoCreatorModalComponent;
}());
export { BoCreatorModalComponent };
if (false) {
    /** @type {?} */
    BoCreatorModalComponent.prototype.parentId;
    /** @type {?} */
    BoCreatorModalComponent.prototype.boConfirmEmitter;
    /** @type {?} */
    BoCreatorModalComponent.prototype.boGroupSelector;
    /**
     * 当前BO节点
     * @type {?}
     */
    BoCreatorModalComponent.prototype.currentBo;
    /**
     * 关键应用集合
     * @type {?}
     */
    BoCreatorModalComponent.prototype.items;
    /**
     * 选择应用节点下的所有服务单元集合
     * @type {?}
     */
    BoCreatorModalComponent.prototype.suList;
    /**
     * 当前选择的SU下的所有BO
     * @type {?}
     */
    BoCreatorModalComponent.prototype.bosInSU;
    /**
     * 所有业务对象节点集合
     * @type {?}
     */
    BoCreatorModalComponent.prototype.data;
    /**
     * 所属微服务
     * @type {?}
     */
    BoCreatorModalComponent.prototype.suCode;
    /** @type {?} */
    BoCreatorModalComponent.prototype.parentBo;
    /** @type {?} */
    BoCreatorModalComponent.prototype.labelLength;
    /** @type {?} */
    BoCreatorModalComponent.prototype.inputLength;
    /**
     * @type {?}
     * @private
     */
    BoCreatorModalComponent.prototype.boCreatorUtils;
    /**
     * @type {?}
     * @private
     */
    BoCreatorModalComponent.prototype.boDtService;
    /**
     * @type {?}
     * @private
     */
    BoCreatorModalComponent.prototype.notifyService;
}
