// @qdp/common/esm5/lib/service/print-integration/print-integrationmanager.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// import 'rxjs/add/operator/map';
import { Injectable, Inject } from '@angular/core';
import { map } from 'rxjs/operators';
import { Subject } from 'rxjs';
import { SessionService } from '@ecp-caf/caf-common';
import { RestfulService } from '../restful-service/restful.service';
import { Server_Host } from '../../entitties/server.host';
import { RtfServices } from '../rtfservices/rtfservices';
var PrintIntegrationManagerService = /** @class */ (function () {
    function PrintIntegrationManagerService(restService, sessionService, host) {
        this.restService = restService;
        this.sessionService = sessionService;
        this.host = host;
        this.uri = '/api/runtime/bcc/v1.0/printintegration/';
        this.printIntegrationInfoChange = new Subject();
        this.printIntegrationInfoList = {};
    }
    /**
     * 获取查询参数映射信息
     * @param schemaId 方案ID
     * @param queryId 查询编号
     */
    /**
     * 获取查询参数映射信息
     * @param {?} schemaId 方案ID
     * @param {?} queryId 查询编号
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    PrintIntegrationManagerService.prototype.getPrintIntegration = /**
     * 获取查询参数映射信息
     * @param {?} schemaId 方案ID
     * @param {?} queryId 查询编号
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (schemaId, queryId, queryRelativeUrl) {
        /** @type {?} */
        var result$ = this.getPrintIntegrationList(queryId, schemaId, '', '', queryRelativeUrl);
        return result$.pipe(map((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value && value[schemaId + queryId]) {
                return value[schemaId + queryId];
            }
            else {
                return null;
            }
        })));
    };
    /**
     * 获取打印设置信息列表
     * @param queryId 查询编号
     * @param schemaId 方案ID
     * @param userId 用户ID
     * @param languageType 语言类型
     */
    /**
     * 获取打印设置信息列表
     * @param {?} queryId 查询编号
     * @param {?} schemaId 方案ID
     * @param {?} userId 用户ID
     * @param {?} languageType 语言类型
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    PrintIntegrationManagerService.prototype.getPrintIntegrationList = /**
     * 获取打印设置信息列表
     * @param {?} queryId 查询编号
     * @param {?} schemaId 方案ID
     * @param {?} userId 用户ID
     * @param {?} languageType 语言类型
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (queryId, schemaId, userId, languageType, queryRelativeUrl) {
        var _this = this;
        /** @type {?} */
        var restUri = this.host + this.uri + 'getprintintegrationlist';
        /** @type {?} */
        var result$ = this.restService.get(restUri, { queryId: queryId, schemaId: schemaId, userId: userId, languageType: languageType }, this.createHeaderSessionId());
        return result$.pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data && data.length) {
                data.forEach((/**
                 * @param {?} element
                 * @return {?}
                 */
                function (element) {
                    element.printintegration = JSON.parse(element.printintegration);
                    _this.printIntegrationInfoList[schemaId + queryId] = element;
                }));
                _this.printIntegrationInfoChange.next({ printIntegrationInfoList: _this.printIntegrationInfoList });
                return _this.printIntegrationInfoList;
            }
            else {
                _this.printIntegrationInfoChange.next({ printIntegrationInfoList: null });
                return null;
            }
        })));
    };
    /**
     * 保存打印设置信息
     * @param element 打印设置
     */
    /**
     * 保存打印设置信息
     * @param {?} element 打印设置
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    PrintIntegrationManagerService.prototype.savePrintIntegration = /**
     * 保存打印设置信息
     * @param {?} element 打印设置
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (element, queryRelativeUrl) {
        var _this = this;
        /** @type {?} */
        var restUri = this.host + this.uri + 'saveprintintegration';
        /** @type {?} */
        var integration = element;
        /** @type {?} */
        var result$ = this.restService.post(restUri, integration, {}, this.createHeaderSessionId());
        return result$.pipe(map((/**
         * @return {?}
         */
        function () {
            element.printintegration = JSON.parse(element.printintegration);
            if (Object.keys(_this.printIntegrationInfoList).indexOf(element.schemaId + element.queryId)) {
                _this.printIntegrationInfoList[element.schemaId + element.queryId] = element;
            }
        })));
    };
    /**
     * 删除打印设置信息
     * @param schemaId 方案ID
     * @param queryId 查询编号
     */
    /**
     * 删除打印设置信息
     * @param {?} schemaId 方案ID
     * @param {?} queryId 查询编号
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    PrintIntegrationManagerService.prototype.deletePrintIntegration = /**
     * 删除打印设置信息
     * @param {?} schemaId 方案ID
     * @param {?} queryId 查询编号
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (schemaId, queryId, queryRelativeUrl) {
        var _this = this;
        /** @type {?} */
        var restUri = this.host + this.uri + 'deleteprintintegration';
        /** @type {?} */
        var result$ = this.restService.delete(restUri, { schemaId: schemaId, queryId: queryId }, this.createHeaderSessionId());
        return result$.pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data === true) {
                _this.printIntegrationInfoList[schemaId + queryId] = null;
            }
            else {
                return false;
            }
        })));
    };
    /**
     * @private
     * @return {?}
     */
    PrintIntegrationManagerService.prototype.createHeaderSessionId = /**
     * @private
     * @return {?}
     */
    function () {
        return RtfServices.createHeaderSessionId(this.sessionService);
    };
    PrintIntegrationManagerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    PrintIntegrationManagerService.ctorParameters = function () { return [
        { type: RestfulService },
        { type: SessionService },
        { type: String, decorators: [{ type: Inject, args: [Server_Host,] }] }
    ]; };
    return PrintIntegrationManagerService;
}());
export { PrintIntegrationManagerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PrintIntegrationManagerService.prototype.uri;
    /** @type {?} */
    PrintIntegrationManagerService.prototype.printIntegrationInfoChange;
    /** @type {?} */
    PrintIntegrationManagerService.prototype.printIntegrationInfoList;
    /**
     * @type {?}
     * @private
     */
    PrintIntegrationManagerService.prototype.restService;
    /**
     * @type {?}
     * @private
     */
    PrintIntegrationManagerService.prototype.sessionService;
    /**
     * @type {?}
     * @private
     */
    PrintIntegrationManagerService.prototype.host;
}
