// @farris/devkit/esm5/lib/effector/dependency_effector.js
import { Inject, Injectable, Injector } from '@angular/core';
import { BindingData } from '../binding-data/index';
import { EntityList } from '../entity/index';
import { NAMESPACE } from '../frame/tokens';
import { Repository } from '../repository/index';
/**
 * 依赖副作用器
 * @description 当表达式计算结果为true时，清空字段值
 * 当value==true且属性当前值不是null时，设置属性值为null
 */
var DependencyEffector = /** @class */ (function () {
    function DependencyEffector(injector, repository, namespace, bindingData) {
        this.injector = injector;
        this.repository = repository;
        this.namespace = namespace;
        this.bindingData = bindingData;
        this.ns = namespace;
    }
    DependencyEffector.prototype.effect = function (path, value, options) {
        if (!options || !options.path) {
            throw new Error('DependencyEffector 需要指定行信息。');
        }
        if (typeof value !== 'boolean') {
            console.warn("DependencyEffector \u4F9D\u8D56\u8868\u8FBE\u5F0F\u8BA1\u7B97\u7ED3\u679C\u5E94\u8BE5\u4E3Atrue/false\uFF0C\u5F53\u524D\u503C\u4E3A\uFF1A" + value);
        }
        var propertyPath = options.path;
        var rowId = propertyPath[0] || this.bindingData.list.currentItem.primaryKeyValue;
        var entity = this.repository.entityCollection.getEntityById(rowId);
        if (rowId && !entity) {
            throw new Error("\u627E\u4E0D\u5230id\uFF1A" + rowId + "\u5BF9\u5E94\u7684\u5B9E\u4F53\uFF01");
        }
        // propertyPath like : [1, child1s, 1.1, child2s, 1.1.1,propName] or [1,udt,propName] or [1,prop]
        var propName = propertyPath.pop();
        var object = entity;
        for (var index = 1; index < propertyPath.length; index++) {
            var propertyName = propertyPath[index];
            if (object instanceof EntityList) {
                object = object.get(propertyName);
            }
            else {
                object = object[propertyName];
            }
        }
        if (object) {
            if (object[propName] !== null && value === true) {
                object[propName] = null;
            }
        }
        else {
            throw new Error("[DependencyEffector] \u627E\u4E0D\u5230\u5B9E\u4F53\u5BF9\u5E94\u7684\u8DEF\u5F84\uFF1A" + propertyPath.push(propName));
        }
    };
    DependencyEffector.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DependencyEffector.ctorParameters = function () { return [
        { type: Injector },
        { type: Repository },
        { type: undefined, decorators: [{ type: Inject, args: [NAMESPACE,] }] },
        { type: BindingData }
    ]; };
    return DependencyEffector;
}());
export { DependencyEffector };
