// @farris/devkit/esm5/lib/effector/relative_effector.js
import { Inject, Injectable, Injector } from '@angular/core';
import { NAMESPACE } from '../frame/index';
import { MESSAGE_SERVICE_TOKEN, NOTIFY_SERVICE_TOKEN } from '../core/index';
var RelativeEffector = /** @class */ (function () {
    function RelativeEffector(injector, messageService, notifyService, namespace) {
        this.injector = injector;
        this.messageService = messageService;
        this.notifyService = notifyService;
        this.namespace = namespace;
        this.ns = namespace;
    }
    RelativeEffector.prototype.effect = function (path, value, options) {
        if (value === true && options.message && this.notifyService) {
            var messageType = options.messageType || 'info';
            this.notifyService[messageType](options.message, { hideTitle: true });
        }
    };
    RelativeEffector.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    RelativeEffector.ctorParameters = function () { return [
        { type: Injector },
        { type: undefined, decorators: [{ type: Inject, args: [MESSAGE_SERVICE_TOKEN,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [NOTIFY_SERVICE_TOKEN,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [NAMESPACE,] }] }
    ]; };
    return RelativeEffector;
}());
export { RelativeEffector };
