// @gsp-svc/cloudprint/esm2015/lib/cloudprint.internalservice.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, ApplicationRef } from '@angular/core';
import { ComponentFactoryResolver } from '@angular/core';
import { PrintEntity } from './entitys/printEntity';
import { NotifyService } from '@farris/ui-notify';
import { Urls } from './utils/cloudprint.Urls';
import { AppService } from '@gsp-sys/rtf-common';
import { PrintFormatService } from './services/printFormatService';
import { WindowService } from '@progress/kendo-angular-dialog';
import { DataValidator } from './utils/cloudprint.datavalidator';
import { BsModalService } from '@farris/ui-modal';
export class CloudprintInternalService {
    /**
     * @param {?} formatSrv
     * @param {?} resolver
     * @param {?} injector
     * @param {?} applicationRef
     * @param {?} notifyService
     * @param {?} appService
     * @param {?} windowService
     * @param {?} modalService
     */
    constructor(formatSrv, resolver, injector, applicationRef, notifyService, appService, windowService, modalService) {
        this.formatSrv = formatSrv;
        this.resolver = resolver;
        this.injector = injector;
        this.applicationRef = applicationRef;
        this.notifyService = notifyService;
        this.appService = appService;
        this.windowService = windowService;
        this.modalService = modalService;
        /**
         * 打印参数实体
         */
        this.printEntity = new PrintEntity();
        // 设置printEntity的默认值
        this.printEntity.appUrl = Urls.CloudPrintBaseUrl;
        this.printEntity.isUseMetaData = true;
        this.printEntity.language = localStorage.getItem("languageCode") || 'zh-CHS';
        this.formatSrv.printEntity = this.printEntity;
    }
    /**
     * 设置PrintEntity参数
     * \@description 如果外界new新的PrintEntity，则需要调用该方法同步service的entity
     *              外界没有new新的PrintEntity，则不需要调用该方法
     * @param {?} entity
     * @return {?}
     */
    set setPrintEntity(entity) {
        this.printEntity = entity;
        // this.designerSrv.printEntity = this.printEntity;
        // this.viewerSrv.printEntity = this.printEntity;
        this.formatSrv.printEntity = this.printEntity;
    }
    /**
     * be数据源创建格式
     * @param {?} beId
     * @return {?}
     */
    createPrintFormatWithBE(beId) {
        return this.formatSrv.createPrintFormatWithBE(beId);
    }
    /**
     * vo数据源创建格式
     * @param {?} voId
     * @return {?}
     */
    createPrintFormatWithVO(voId) {
        return this.formatSrv.createPrintFormatWithVO(voId);
    }
    /**
     * 根据数据源id获取格式
     * @param {?} sourceId
     * @return {?}
     */
    getPrintFormatByDataSrouceId(sourceId) {
        return this.formatSrv.getFormatBySourceId(sourceId);
    }
    /**
     * 根据数据源和维度获取格式
     * @param {?} billCategoryId
     * @param {?} sourceId
     * @param {?} firstDimension
     * @param {?} secondDimension
     * @return {?}
     */
    getPrintFormatByDataSrouceAndDimension(billCategoryId, sourceId, firstDimension, secondDimension) {
        return this.formatSrv.getPrintFormatByDataSrouceAndDimension(billCategoryId, sourceId, firstDimension, secondDimension);
    }
    /**
     * 根据业务对象id获取数据源
     * @param {?} boId
     * @return {?}
     */
    getDataSourceByBOId(boId) {
        return this.formatSrv.getDataSourceByBOId(boId);
    }
    /**
     * 根据格式id获取格式
     * @param {?} formatId
     * @return {?}
     */
    getFormatByFormatId(formatId) {
        return this.formatSrv.getFormatByFormatId(formatId);
    }
    /**
     * 根据已有的格式id和数据源id，实时获取数据源结构后组成新的格式
     * @param {?} formatId
     * @return {?}
     */
    getNewSourceFormatByFormatId(formatId) {
        return this.formatSrv.getNewSourceFormatByFormatId(formatId);
    }
    /**
     * 根据已有的格式id和数据源id，语言，实时获取数据源结构后组成新的格式
     * @param {?} formatId
     * @param {?} lang
     * @return {?}
     */
    getNewSourceFormatByFormatIdAndLang(formatId, lang) {
        return this.formatSrv.getNewSourceFormatByFormatIdAndLang(formatId, lang);
    }
    /**
     * 保存格式
     * @param {?} data
     * @return {?}
     */
    save(data) {
        return this.formatSrv.save(data);
    }
    /**
     * 保存格式
     * @param {?} data
     * @return {?}
     */
    saveWithRes(data) {
        return this.formatSrv.saveWithRes(data);
    }
    /**
     * 删除格式
     * @param {?} formatId
     * @return {?}
     */
    remove(formatId) {
        return this.formatSrv.remove(formatId);
    }
    /**
     * 根据格式id获取格式
     * @param {?} id
     * @return {?}
     */
    getFormatById(id) {
        if (DataValidator.checkAppUrl(this.printEntity.appUrl) == false) {
            this.notifyService.info('未配置正确的appUrl');
            return;
        }
        return this.formatSrv.loadFormat(id, 'CloudPrint');
    }
}
CloudprintInternalService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CloudprintInternalService.ctorParameters = () => [
    { type: PrintFormatService },
    { type: ComponentFactoryResolver },
    { type: Injector },
    { type: ApplicationRef },
    { type: NotifyService },
    { type: AppService },
    { type: WindowService },
    { type: BsModalService }
];
if (false) {
    /**
     * 打印参数实体
     * @type {?}
     */
    CloudprintInternalService.prototype.printEntity;
    /**
     * @type {?}
     * @private
     */
    CloudprintInternalService.prototype.formatSrv;
    /**
     * @type {?}
     * @private
     */
    CloudprintInternalService.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    CloudprintInternalService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    CloudprintInternalService.prototype.applicationRef;
    /**
     * @type {?}
     * @private
     */
    CloudprintInternalService.prototype.notifyService;
    /**
     * @type {?}
     * @private
     */
    CloudprintInternalService.prototype.appService;
    /**
     * @type {?}
     * @private
     */
    CloudprintInternalService.prototype.windowService;
    /**
     * @type {?}
     * @private
     */
    CloudprintInternalService.prototype.modalService;
}
