// @gsp-svc/cloudprint/esm2015/lib/cloudprint.attachmentservice.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, Injector, ApplicationRef, LOCALE_ID } from '@angular/core';
import { ComponentFactoryResolver } from '@angular/core';
import { PrintEntity } from './entitys/printEntity';
import { NotifyService } from '@farris/ui-notify';
import { Urls } from './utils/cloudprint.Urls';
import { AppService } from '@gsp-sys/rtf-common';
import { PrintFormatService } from './services/printFormatService';
import { QueryType } from './entitys/enums/querytype';
import { PrintType } from './entitys/enums/printtype';
import { DataValidator } from './utils/cloudprint.datavalidator';
import { FormatlistjobComponent } from './cloudprint-formatlistjob/formatlistjob.component';
import { BsModalService } from '@farris/ui-modal';
import { LocalLangPipe } from './local.pipe';
import { HttpClient } from '@angular/common/http';
import { Config } from './entitys/config';
/**
 * 附件打印的服务
 */
export class CloudPrintAttachmentService {
    /**
     * @param {?} formatSrv
     * @param {?} resolver
     * @param {?} injector
     * @param {?} applicationRef
     * @param {?} notifyService
     * @param {?} appService
     * @param {?} modalService
     * @param {?} httpClient
     */
    constructor(formatSrv, resolver, injector, applicationRef, notifyService, appService, modalService, httpClient) {
        this.formatSrv = formatSrv;
        this.resolver = resolver;
        this.injector = injector;
        this.applicationRef = applicationRef;
        this.notifyService = notifyService;
        this.appService = appService;
        this.modalService = modalService;
        this.httpClient = httpClient;
        /**
         * 打印参数实体
         */
        this.printEntity = new PrintEntity();
        // 设置printEntity的默认值
        this.printEntity.appUrl = Urls.CloudPrintBaseUrl;
        this.printEntity.isUseMetaData = true;
        this.printEntity.language = localStorage.getItem("languageCode") || 'zh-CHS';
        this.formatSrv.printEntity = this.printEntity;
        if (this.injector) {
            this.modalService = this.injector.get(BsModalService);
            this.localeid = localStorage.getItem("languageCode") || this.injector.get(LOCALE_ID);
            this.localpipe = new LocalLangPipe(this.localeid);
        }
    }
    /**
     * 设置PrintEntity参数
     * \@description 如果外界new新的PrintEntity，则需要调用该方法同步service的entity
     *              外界没有new新的PrintEntity，则不需要调用该方法
     * @param {?} entity
     * @return {?}
     */
    set setPrintEntity(entity) {
        this.printEntity = entity;
        // this.designerSrv.printEntity = this.printEntity;
        // this.viewerSrv.printEntity = this.printEntity;
        this.formatSrv.printEntity = this.printEntity;
    }
    /**
     * @param {?} attachmentoptions
     * @param {?} options
     * @param {?=} su
     * @param {?=} param
     * @return {?}
     */
    printAttachment(attachmentoptions, options, su = null, param = null) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (DataValidator.checkAppUrl(this.printEntity.appUrl) == false) {
                this.notifyService.info(this.localpipe.transform('appurlerror'));
                return;
            }
            if (attachmentoptions.queryType == undefined) { //兼容以前的逻辑
                attachmentoptions.queryType = QueryType.Unknown;
            }
            if (options.printType == undefined) { //兼容以前的逻辑这里默认为表单打印
                options.printType = PrintType.Attachment;
            }
            /** @type {?} */
            const that = this;
            return this.formatSrv.getPrintDevice().then((/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                /** @type {?} */
                const device = data || [];
                /** @type {?} */
                const factory = that.resolver.resolveComponentFactory(FormatlistjobComponent);
                /** @type {?} */
                const printsetting = factory.create(that.injector);
                printsetting.instance.boxlist = device;
                printsetting.instance.formatId = '';
                printsetting.instance.isShowFormat = false;
                printsetting.instance.printType = 1;
                printsetting.instance.localepip = that.localpipe;
                /** @type {?} */
                const dlg = this.modalService.show(printsetting, {
                    title: this.localpipe.transform('printset'),
                    width: 800,
                    height: 500,
                    showButtons: true,
                    showMaxButton: false,
                    buttons: [
                        { text: this.localpipe.transform('print'), cls: 'btn btn-primary', handle: (/**
                             * @return {?}
                             */
                            () => {
                                printsetting.instance.printJob();
                            }) },
                        { text: this.localpipe.transform('cancel'), cls: 'btn btn-secondary', handle: (/**
                             * @return {?}
                             */
                            () => {
                                dlg.close();
                            }) }
                    ],
                });
                //订阅了弹出的格式选择列表的关闭事件
                printsetting.instance.print.subscribe((/**
                 * @param {?} data
                 * @return {?}
                 */
                (data) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                    dlg.close();
                    options.printSetting = data.printSetting;
                    options.printerName = data.printter;
                    options.DeviceId = data.boxId;
                    options.printJob = true;
                    return yield this.printAttachmentJob(attachmentoptions, options);
                })));
            }));
        });
    }
    /**
     * @private
     * @param {?} attachmentoptions
     * @param {?} options
     * @return {?}
     */
    printAttachmentJob(attachmentoptions, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /** @type {?} */
            const url = this.printEntity.appUrl + Urls.PrintAttachment + 'su=' + Config.PRINT_SU;
            ;
            /** @type {?} */
            const httpoptions = {};
            httpoptions['headers'] = {
                SessionId: this.printEntity.sessionId || '',
                'Content-Type': 'application/json'
            };
            /** @type {?} */
            const data = {
                attachmentOptions: attachmentoptions,
                options: options,
            };
            return yield this.httpClient.post(url, data, httpoptions).toPromise()
                .then((/**
             * @return {?}
             */
            () => {
                this.notifyService.info(this.localpipe.transform('sendprintjob'));
                return this.localpipe.transform('sendprintjob');
            }))
                .catch((/**
             * @param {?} err
             * @return {?}
             */
            (err) => {
                this.notifyService.warning(err.error.Message);
                return err.error.Message;
            }));
        });
    }
}
CloudPrintAttachmentService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CloudPrintAttachmentService.ctorParameters = () => [
    { type: PrintFormatService },
    { type: ComponentFactoryResolver },
    { type: Injector },
    { type: ApplicationRef },
    { type: NotifyService },
    { type: AppService },
    { type: BsModalService },
    { type: HttpClient }
];
if (false) {
    /**
     * 打印参数实体
     * @type {?}
     */
    CloudPrintAttachmentService.prototype.printEntity;
    /** @type {?} */
    CloudPrintAttachmentService.prototype.localeid;
    /** @type {?} */
    CloudPrintAttachmentService.prototype.localpipe;
    /**
     * @type {?}
     * @private
     */
    CloudPrintAttachmentService.prototype.formatSrv;
    /**
     * @type {?}
     * @private
     */
    CloudPrintAttachmentService.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    CloudPrintAttachmentService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    CloudPrintAttachmentService.prototype.applicationRef;
    /**
     * @type {?}
     * @private
     */
    CloudPrintAttachmentService.prototype.notifyService;
    /**
     * @type {?}
     * @private
     */
    CloudPrintAttachmentService.prototype.appService;
    /**
     * @type {?}
     * @private
     */
    CloudPrintAttachmentService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    CloudPrintAttachmentService.prototype.httpClient;
}
