// @qdp/localize/esm5/lib/localize.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, LOCALE_ID, Inject } from '@angular/core';
import { CommonUtils } from '@farris/ui-common';
import { QDP_LOCALIZE, DEFAULT_LANGUAGE } from './i18n/index';
import * as i0 from "@angular/core";
import * as i1 from "@farris/ui-common";
var LocalizeService = /** @class */ (function () {
    function LocalizeService(localeId, utils) {
        this.localeId = localeId;
        this.utils = utils;
        if (!localeId) {
            localeId = DEFAULT_LANGUAGE;
            this.language = 'zh-CHS';
        }
        else {
            this.language = localeId;
        }
        if (!QDP_LOCALIZE[localeId.toLowerCase()]) {
            localeId = DEFAULT_LANGUAGE;
        }
        this.languageData = QDP_LOCALIZE[localeId.toLowerCase()];
    }
    /**
     * @param {?} propertyName
     * @return {?}
     */
    LocalizeService.prototype.getValue = /**
     * @param {?} propertyName
     * @return {?}
     */
    function (propertyName) {
        /** @type {?} */
        var val = this.utils.getValue(propertyName, this.languageData);
        if (val) {
            return val;
        }
        else {
            console.warn("Can't find language Data for the " + propertyName);
        }
    };
    LocalizeService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    LocalizeService.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: CommonUtils }
    ]; };
    /** @nocollapse */ LocalizeService.ngInjectableDef = i0.defineInjectable({ factory: function LocalizeService_Factory() { return new LocalizeService(i0.inject(i0.LOCALE_ID), i0.inject(i1.CommonUtils)); }, token: LocalizeService, providedIn: "root" });
    return LocalizeService;
}());
export { LocalizeService };
if (false) {
    /** @type {?} */
    LocalizeService.prototype.languageData;
    /** @type {?} */
    LocalizeService.prototype.language;
    /** @type {?} */
    LocalizeService.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    LocalizeService.prototype.utils;
}
