// @gspwidget/widget-devkit/esm5/lib/widget-registry/widget-event.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { curLang, FILTERS_WIDGET_FILTERS_PROP_NAME, FILTERS_WIDGET_NAME } from '@gspwidget/util';
import { WidgetEventMap } from './widget-event.decorator';
import { WidgetRegistry } from './widget-registry';
import * as i0 from "@angular/core";
var WidgetEventService = /** @class */ (function () {
    function WidgetEventService() {
    }
    /**
     * @param {?} widgetName
     * @param {?} pvs
     * @return {?}
     */
    WidgetEventService.prototype.getEventsByWidgetName = /**
     * @param {?} widgetName
     * @param {?} pvs
     * @return {?}
     */
    function (widgetName, pvs) {
        /** @type {?} */
        var ctorName = WidgetRegistry.getWidgetByName(widgetName).ctor.name;
        /** @type {?} */
        var widgetEvents = WidgetEventMap.get(ctorName) || [];
        if (widgetName === FILTERS_WIDGET_NAME) {
            return pvs[FILTERS_WIDGET_FILTERS_PROP_NAME].map((/**
             * @param {?} fc
             * @return {?}
             */
            function (fc) { return ({
                name: fc.i18nName[curLang],
                propName: fc.i18nName[curLang],
            }); }));
        }
        // 组装
        return tslib_1.__spread(widgetEvents, test);
    };
    WidgetEventService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    WidgetEventService.ctorParameters = function () { return []; };
    /** @nocollapse */ WidgetEventService.ngInjectableDef = i0.defineInjectable({ factory: function WidgetEventService_Factory() { return new WidgetEventService(); }, token: WidgetEventService, providedIn: "root" });
    return WidgetEventService;
}());
export { WidgetEventService };
/** @type {?} */
var test = [{
        name: 'click bar'
    }, {
        name: 'event 2'
    }, {
        name: 'event 3'
    }];
