// @gsp-wf/wf-process-editor/esm5/lib/views/execute-forms/execute-forms.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, HostBinding, Output, EventEmitter, Input, ViewChild, TemplateRef, ElementRef, ComponentFactoryResolver, Injector, ChangeDetectorRef, Optional, Inject, LOCALE_ID } from '@angular/core';
import { of } from 'rxjs/observable/of';
import { FormBuilder } from '@angular/forms';
import { ExecuteFormsUIState } from './execute-forms.uistate';
import { ExecuteFormsViewModel } from './execute-forms.viewmodel';
import { ExecuteFormsRepository } from './execute-forms.repository';
import { ExecuteFormModel, ActualParameter } from './entities/execute-form.entity';
import { BsModalService } from '@farris/ui-modal';
import { AddExecuteFormComponent } from './add-execute-form/add-execute-form.component';
import { FormConfigContainerComponent } from './form-config-container/form-config-container.component';
import { ParameterSettingComponent } from './parameter-setting/parameter-setting.component';
import { switchMap } from 'rxjs/operators';
import { ProcessDesignerUIState } from '../../app/process-designer.uistate';
import { MessagerService } from '@farris/ui-messager';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../i18n/pipe/wf-process-editor-locale.pipe';
var ExecuteFormsComponent = /** @class */ (function () {
    // *******************************************************
    function ExecuteFormsComponent(formBuilder, uistate, viewmodel, resolver, injector, modalService, designerState, message, cdr, localeId, resources, localePipe) {
        this.formBuilder = formBuilder;
        this.uistate = uistate;
        this.viewmodel = viewmodel;
        this.resolver = resolver;
        this.injector = injector;
        this.modalService = modalService;
        this.designerState = designerState;
        this.message = message;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.modalConfig = {
            title: this.localePipe.transform('component.executeForms.executeForms'),
            width: 760,
            height: 532,
            showButtons: true
        };
        this.cls = 'farris-main-area flex-column';
        this.style = '#FFF';
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.editorParams = {};
        this.tab = 'web';
        this.groupIcon = '<i class="f-icon f-icon-lookup"></i>';
        this.cols = [];
        this.showMethodTab = true;
        this.allParameters = new Array();
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
        this.form = new ExecuteFormModel();
        this.allParameters = this.getVariableList();
    }
    Object.defineProperty(ExecuteFormsComponent.prototype, "value", {
        set: /**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            var _this = this;
            if (item) {
                /** @type {?} */
                var paramArray_1 = new Array();
                item.ActualParameters.forEach((/**
                 * @param {?} a
                 * @return {?}
                 */
                function (a) {
                    /** @type {?} */
                    var para = new ActualParameter();
                    para.Name = a.Name;
                    para.Tag = a.Tag ? a.Tag : _this.initVariableType(a.Value);
                    para.Value = a.Value;
                    paramArray_1.push(para);
                }));
                this.form.ActualParameters = paramArray_1;
                this.form.Id = item.Id;
                this.form.Name = item.Name;
                this.form.ComponentId = item.ComponentId;
                this.form.formFormatId = item.formFormatId;
                this.form.formatConfigId = item.formatConfigId;
                this.form.FormType = item.FormType;
                this.form.TerminalType = item.TerminalType;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ExecuteFormsComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    ExecuteFormsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.location = this.editorParams.location;
        this.tab = this.editorParams.tab;
        if (this.location === 'userActivity') {
            this.showMethodTab = true;
        }
        else {
            this.showMethodTab = false;
        }
    };
    /**
     * @return {?}
     */
    ExecuteFormsComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var data = new ExecuteFormModel();
        this.getForm().subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            var e_1, _a;
            if (!re.result) {
                return;
            }
            data = re.form;
            if (_this.validator(data)) {
                /** @type {?} */
                var parameters = null;
                // 如果存在变量类型参数在流程变量上不存在，需要同步流程变量
                if (data && data.ActualParameters && data.ActualParameters.length > 0) {
                    /** @type {?} */
                    var formalParameters = _this.designerState.flowFormInfo.formalParameters.filter((/**
                     * @param {?} p
                     * @return {?}
                     */
                    function (p) { return p.dataType !== 'CDTSchema'; }));
                    var _loop_1 = function (parameter) {
                        if (parameter.Tag === 'variable' && parameter.Value && !_this.allParameters.some((/**
                         * @param {?} p
                         * @return {?}
                         */
                        function (p) { return p.key === parameter.Value; })) && formalParameters.some((/**
                         * @param {?} p
                         * @return {?}
                         */
                        function (p) { return p.code === parameter.Value; }))) {
                            parameters = formalParameters;
                            return "break";
                        }
                    };
                    try {
                        for (var _b = tslib_1.__values(data.ActualParameters), _c = _b.next(); !_c.done; _c = _b.next()) {
                            var parameter = _c.value;
                            var state_1 = _loop_1(parameter);
                            if (state_1 === "break")
                                break;
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                }
                /** @type {?} */
                var object = { value: data, parameters: parameters };
                _this.submitModal.emit(object);
            }
        }));
    };
    /**
     * @return {?}
     */
    ExecuteFormsComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeModal.emit();
    };
    /**
     * @param {?} form
     * @return {?}
     */
    ExecuteFormsComponent.prototype.validator = /**
     * @param {?} form
     * @return {?}
     */
    function (form) {
        var e_2, _a;
        if (form && form.ActualParameters && form.ActualParameters.length > 0) {
            /** @type {?} */
            var formalParameters = this.designerState.flowFormInfo.formalParameters.filter((/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return p.dataType !== 'CDTSchema'; }));
            var _loop_2 = function (parameter) {
                if (parameter.Tag === 'variable' && parameter.Value && !this_1.allParameters.some((/**
                 * @param {?} p
                 * @return {?}
                 */
                function (p) { return p.key === parameter.Value; })) && !formalParameters.some((/**
                 * @param {?} p
                 * @return {?}
                 */
                function (p) { return p.code === parameter.Value; }))) {
                    this_1.message.error(this_1.localePipe.transform('component.executeForms.parameters') + parameter.Name + this_1.localePipe.transform('component.executeForms.typeAsvariable'));
                    return { value: false };
                }
            };
            var this_1 = this;
            try {
                for (var _b = tslib_1.__values(form.ActualParameters), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var parameter = _c.value;
                    var state_2 = _loop_2(parameter);
                    if (typeof state_2 === "object")
                        return state_2.value;
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        return true;
    };
    // tabChange(data) {
    //   this.tab = data.nextId;
    // }
    // editCSS() {
    //   this.configSection.nativeElement.querySelector('div.f-section-header').classList.add('mb-0');
    // }
    // tabChange(data) {
    //   this.tab = data.nextId;
    // }
    // editCSS() {
    //   this.configSection.nativeElement.querySelector('div.f-section-header').classList.add('mb-0');
    // }
    /**
     * @private
     * @return {?}
     */
    ExecuteFormsComponent.prototype.getForm = 
    // tabChange(data) {
    //   this.tab = data.nextId;
    // }
    // editCSS() {
    //   this.configSection.nativeElement.querySelector('div.f-section-header').classList.add('mb-0');
    // }
    /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.form.ComponentId) {
            return of({ result: true, form: this.form });
        }
        else if (this.form.formFormatId) {
            return this.configContainer.submit().pipe(switchMap((/**
             * @param {?} re
             * @return {?}
             */
            function (re) {
                if (re.result) {
                    _this.form.formatConfigId = re.configId;
                    return of({ result: true, form: _this.form });
                }
                else {
                    return of({ result: false, form: _this.form });
                }
            })));
        }
        else {
            return of({ result: true, form: null });
        }
    };
    /**
     * 参数设置
     */
    /**
     * 参数设置
     * @return {?}
     */
    ExecuteFormsComponent.prototype.parametersSet = /**
     * 参数设置
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var compFactory = this.resolver.resolveComponentFactory(ParameterSettingComponent);
        /** @type {?} */
        var compRef = compFactory.create(this.injector);
        compRef.instance.mode = 'bpmn';
        compRef.instance.allParameters = this.allParameters;
        if (this.tab === 'web') {
            if (this.form.ComponentId) {
                compRef.instance.showFormType = true;
                compRef.instance.formType = this.form.FormType;
            }
            compRef.instance.parameters = tslib_1.__spread(this.form.ActualParameters.map((/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return Object.assign({}, p); })));
        }
        else {
            compRef.instance.parameters = tslib_1.__spread(this.form.ActualParameters.map((/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return Object.assign({}, p); })));
        }
        /** @type {?} */
        var modalConfig = {
            title: this.localePipe.transform('component.executeForms.parameterSettings'),
            width: 620,
            height: 400,
            showButtons: true,
            buttons: compRef.instance.modalFooter
        };
        /** @type {?} */
        var dialog = this.modalService.show(compRef, modalConfig);
        compRef.instance.submit.subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            if (_this.tab === 'web') {
                _this.form.ActualParameters = re.parameters;
                _this.form.FormType = re.formType;
            }
            else {
                _this.form.ActualParameters = re.parameters;
            }
            dialog.close();
        }));
        compRef.instance.cancel.subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            dialog.close();
        }));
    };
    /**
     * 选择表单定义
     */
    /**
     * 选择表单定义
     * @return {?}
     */
    ExecuteFormsComponent.prototype.selectExecuteForm = /**
     * 选择表单定义
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var compFactory = this.resolver.resolveComponentFactory(AddExecuteFormComponent);
        /** @type {?} */
        var compRef = compFactory.create(this.injector);
        compRef.instance.tab = this.tab;
        /** @type {?} */
        var modalConfig = {
            title: this.localePipe.transform('component.executeForms.formDefinitionList'),
            width: 650,
            height: 400,
            showButtons: true,
            buttons: compRef.instance.modalFooter
        };
        /** @type {?} */
        var dialog = this.modalService.show(compRef, modalConfig);
        compRef.instance.submit.subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            _this.form = re;
            dialog.close();
        }));
        compRef.instance.cancel.subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            dialog.close();
        }));
    };
    /**
     * @return {?}
     */
    ExecuteFormsComponent.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.form = new ExecuteFormModel();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ExecuteFormsComponent.prototype.initVariableType = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value) {
            if (value.length >= 8 && value.slice(0, 8) === "{\"expr\":") {
                return 'expression';
            }
            else {
                return 'constant';
            }
        }
        else {
            return 'variable';
        }
    };
    /**
     * @private
     * @return {?}
     */
    ExecuteFormsComponent.prototype.getVariableList = /**
     * @private
     * @return {?}
     */
    function () {
        var e_3, _a, e_4, _b, e_5, _c, e_6, _d, e_7, _e;
        /** @type {?} */
        var context = new Array();
        if (this.designerState.procInstContext.length > 0) {
            try {
                for (var _f = tslib_1.__values(this.designerState.procInstContext), _g = _f.next(); !_g.done; _g = _f.next()) {
                    var variable = _g.value;
                    context.push(variable);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_g && !_g.done && (_a = _f.return)) _a.call(_f);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        if (this.designerState.subActivityContext.length > 0) {
            var _loop_3 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _h = tslib_1.__values(this.designerState.subActivityContext), _j = _h.next(); !_j.done; _j = _h.next()) {
                    var variable = _j.value;
                    _loop_3(variable);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_j && !_j.done && (_b = _h.return)) _b.call(_h);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
        if (this.designerState.formalParameterContext.length > 0) {
            var _loop_4 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _k = tslib_1.__values(this.designerState.formalParameterContext), _l = _k.next(); !_l.done; _l = _k.next()) {
                    var variable = _l.value;
                    _loop_4(variable);
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (_l && !_l.done && (_c = _k.return)) _c.call(_k);
                }
                finally { if (e_5) throw e_5.error; }
            }
        }
        if (this.designerState.participantContext.length > 0) {
            var _loop_5 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _m = tslib_1.__values(this.designerState.participantContext), _o = _m.next(); !_o.done; _o = _m.next()) {
                    var variable = _o.value;
                    _loop_5(variable);
                }
            }
            catch (e_6_1) { e_6 = { error: e_6_1 }; }
            finally {
                try {
                    if (_o && !_o.done && (_d = _m.return)) _d.call(_m);
                }
                finally { if (e_6) throw e_6.error; }
            }
        }
        if (this.designerState.processVariableContext.length > 0) {
            var _loop_6 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _p = tslib_1.__values(this.designerState.processVariableContext), _q = _p.next(); !_q.done; _q = _p.next()) {
                    var variable = _q.value;
                    _loop_6(variable);
                }
            }
            catch (e_7_1) { e_7 = { error: e_7_1 }; }
            finally {
                try {
                    if (_q && !_q.done && (_e = _p.return)) _e.call(_p);
                }
                finally { if (e_7) throw e_7.error; }
            }
        }
        return context;
    };
    ExecuteFormsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-execute-forms',
                    template: "\r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm\" (click)=\"clickCancel()\">{{'component.executeForms.clickCancel'|wfLocale}}</button>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">{{'component.executeForms.clickOK'|wfLocale}}</button>\r\n</ng-template>\r\n\r\n<ng-template #empty>\r\n  <div class=\"f-utils-fill-flex-column h-100\">\r\n    <div class=\"f-datagrid-norecords\" style=\"height: 100%;\">\r\n      <div class=\"f-datagrid-norecords-content w-100\">{{'component.executeForms.noForm'|wfLocale}}</div>\r\n    </div>\r\n  </div>\r\n</ng-template>\r\n\r\n<div class=\"f-page f-page-is-managelist\">\r\n  <div class=\"f-page-main mx-3\" style=\"border-radius:16px;border: 1.5px solid #F1F3F7;\">\r\n    <div class=\"farris-form farris-form-controls-inline farris-form-auto f-form-lable-auto\">\r\n      <div class=\"f-section-formgroup-inputs\">\r\n        <div class=\"col-12 col-md-5 px-0 my-2\" style=\"margin-left: 50px;\">\r\n          <div class=\"farris-group-wrap\">\r\n            <div class=\"form-group farris-form-group mb-0\">\r\n              <label class=\"col-form-label\">\r\n                <span class=\"farris-label-text\">{{'component.executeForms.form'|wfLocale}}</span>\r\n              </label>\r\n              <input-group class=\"form-control\" style=\"width: 100%;padding:0;border:0;\" [enableClear]=\"true\"\r\n                [showButtonWhenDisabled]=\"false\" [(ngModel)]=\"form.Name\" name=\"form\" placeholder=\"{{'component.executeForms.placeholder'|wfLocale}}\"\r\n                [editable]=\"false\" (clickHandle)=\"selectExecuteForm()\" (clear)=\"clear()\" [groupText]=\"groupIcon\">\r\n              </input-group>\r\n              <div class=\"farris-feedback valid-feedback\"></div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n        <div class=\"col-12 col-md-2 my-2 px-0\">\r\n          <div class=\"farris-group-wrap\" style=\"padding-top: 2px;\">\r\n            <div class=\"form-group farris-form-group mb-0\">\r\n              <div class=\"farris-input-wrap\">\r\n                <farris-button [disabled]=\"!form.ComponentId && !form.formFormatId\" [type]=\"'link'\" (click)=\"parametersSet()\">\r\n                  {{'component.executeForms.parametersSet'|wfLocale}}\r\n                </farris-button>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n    <div class=\"h-100 mx-3\">\r\n    <lib-form-config-container #configContainer\r\n      [formFormatId]=\"form.formFormatId\" [formatConfigId]=\"form.formatConfigId\" [showMethodTab]=\"showMethodTab\"\r\n      class=\"h-100\"></lib-form-config-container>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n\r\n\r\n\r\n",
                    providers: [
                        ExecuteFormsUIState, ExecuteFormsViewModel, ExecuteFormsRepository,
                        WfLocalePipe,
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ExecuteFormsComponent.ctorParameters = function () { return [
        { type: FormBuilder },
        { type: ExecuteFormsUIState },
        { type: ExecuteFormsViewModel },
        { type: ComponentFactoryResolver },
        { type: Injector },
        { type: BsModalService },
        { type: ProcessDesignerUIState },
        { type: MessagerService },
        { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
        { type: WfLocalePipe, decorators: [{ type: Optional }] }
    ]; };
    ExecuteFormsComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        style: [{ type: HostBinding, args: ['style.background',] }],
        closeModal: [{ type: Output }],
        submitModal: [{ type: Output }],
        value: [{ type: Input }],
        editorParams: [{ type: Input }],
        modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }],
        configContainer: [{ type: ViewChild, args: ['configContainer',] }],
        configSection: [{ type: ViewChild, args: ['configSection', { read: ElementRef },] }]
    };
    return ExecuteFormsComponent;
}());
export { ExecuteFormsComponent };
if (false) {
    /** @type {?} */
    ExecuteFormsComponent.prototype.modalConfig;
    /** @type {?} */
    ExecuteFormsComponent.prototype.cls;
    /** @type {?} */
    ExecuteFormsComponent.prototype.style;
    /** @type {?} */
    ExecuteFormsComponent.prototype.closeModal;
    /** @type {?} */
    ExecuteFormsComponent.prototype.submitModal;
    /** @type {?} */
    ExecuteFormsComponent.prototype.editorParams;
    /** @type {?} */
    ExecuteFormsComponent.prototype.modalFooter;
    /** @type {?} */
    ExecuteFormsComponent.prototype.configContainer;
    /** @type {?} */
    ExecuteFormsComponent.prototype.configSection;
    /** @type {?} */
    ExecuteFormsComponent.prototype.form;
    /** @type {?} */
    ExecuteFormsComponent.prototype.tab;
    /** @type {?} */
    ExecuteFormsComponent.prototype.groupIcon;
    /** @type {?} */
    ExecuteFormsComponent.prototype.cols;
    /** @type {?} */
    ExecuteFormsComponent.prototype.location;
    /** @type {?} */
    ExecuteFormsComponent.prototype.showMethodTab;
    /** @type {?} */
    ExecuteFormsComponent.prototype.allParameters;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.formBuilder;
    /** @type {?} */
    ExecuteFormsComponent.prototype.uistate;
    /** @type {?} */
    ExecuteFormsComponent.prototype.viewmodel;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.designerState;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.message;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.localePipe;
}
var ExecuteFormsConverter = /** @class */ (function () {
    function ExecuteFormsConverter(params, localePipe) {
        this.localePipe = localePipe;
        this.params = params;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    ExecuteFormsConverter.prototype.convertTo = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data) {
            return data.Name;
        }
        else {
            if (this.params.location === 'process') {
                return this.localePipe.transform('component.executeForms.notSet');
            }
            else {
                return this.localePipe.transform('component.executeForms.default');
            }
        }
    };
    return ExecuteFormsConverter;
}());
export { ExecuteFormsConverter };
if (false) {
    /** @type {?} */
    ExecuteFormsConverter.prototype.params;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsConverter.prototype.localePipe;
}
