// @farris/ide-devkit/esm2015/lib/components/ng2-right-click-menu/sh-attach-menu.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input } from '@angular/core';
import { ShContextMenuService } from './sh-context-menu.service';
import { ShContextMenuComponent } from './sh-context-menu.component';
import { fromEvent } from 'rxjs/observable/fromEvent';
import { merge } from 'rxjs/observable/merge';
export class ShAttachMenuDirective {
    /**
     * @param {?} ctxService
     * @param {?} elm
     */
    constructor(ctxService, elm) {
        this.ctxService = ctxService;
        this.elm = elm;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.setupEvents();
    }
    /**
     * @private
     * @return {?}
     */
    setupEvents() {
        /** @type {?} */
        const observables = [];
        if (!this.triggers) {
            observables.push(fromEvent(this.elm.nativeElement, 'contextmenu'));
        }
        else {
            this.triggers.forEach((t) => {
                observables.push(fromEvent(this.elm.nativeElement, t));
            });
        }
        this.sub = merge(...observables).subscribe(this.openMenu.bind(this));
    }
    /**
     * @param {?} event
     * @return {?}
     */
    openMenu(event) {
        this.ctxService.openMenu({
            menu: this.menu,
            mouseEvent: event,
            targetElement: this.elm,
            data: this.data
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.sub.unsubscribe();
    }
}
ShAttachMenuDirective.decorators = [
    { type: Directive, args: [{
                selector: '[shAttachMenu]'
            },] }
];
/** @nocollapse */
ShAttachMenuDirective.ctorParameters = () => [
    { type: ShContextMenuService },
    { type: ElementRef }
];
ShAttachMenuDirective.propDecorators = {
    menu: [{ type: Input, args: ['shAttachMenu',] }],
    triggers: [{ type: Input, args: ['shMenuTriggers',] }],
    data: [{ type: Input, args: ['shMenuData',] }]
};
if (false) {
    /** @type {?} */
    ShAttachMenuDirective.prototype.menu;
    /** @type {?} */
    ShAttachMenuDirective.prototype.triggers;
    /** @type {?} */
    ShAttachMenuDirective.prototype.data;
    /** @type {?} */
    ShAttachMenuDirective.prototype.sub;
    /**
     * @type {?}
     * @private
     */
    ShAttachMenuDirective.prototype.ctxService;
    /**
     * @type {?}
     * @private
     */
    ShAttachMenuDirective.prototype.elm;
}
