// @farris/ide-devkit/esm2015/lib/components/ng2-right-click-menu/sh-context-menu.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ShContextMenuComponent } from './sh-context-menu.component';
import { ShAttachMenuDirective } from './sh-attach-menu.directive';
import { ShContextMenuService } from './sh-context-menu.service';
import { ShContextMenuItemDirective } from './sh-context-menu-item.directive';
import { OverlayModule } from '@angular/cdk/overlay';
export class ShContextMenuModule {
}
ShContextMenuModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    ShAttachMenuDirective,
                    ShContextMenuComponent,
                    ShContextMenuItemDirective
                ],
                exports: [
                    ShAttachMenuDirective,
                    ShContextMenuComponent,
                    ShContextMenuItemDirective
                ],
                providers: [
                    ShContextMenuService
                ],
                imports: [
                    CommonModule,
                    OverlayModule
                ],
                entryComponents: [
                    ShContextMenuComponent
                ]
            },] }
];
