// @farris/ide-devkit/esm2015/lib/components/ng2-right-click-menu/sh-context-menu-item.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, Input, Optional, Output, TemplateRef } from '@angular/core';
export class MenuItemContext {
    constructor() {
        this.$implicit = {};
    }
}
if (false) {
    /** @type {?} */
    MenuItemContext.prototype.$implicit;
}
export class ShContextMenuItemDirective {
    /**
     * @param {?} template
     */
    constructor(template) {
        this.template = template;
        this.closeOnClick = true;
        this.click = new EventEmitter();
        this.context = new MenuItemContext();
    }
    /**
     * @return {?}
     */
    setNotActive() {
        this._active = false;
        if (this.subMenu) {
            this.subMenu.setNotActive();
        }
    }
    /**
     * @return {?}
     */
    setActive() {
        this._active = true;
    }
}
ShContextMenuItemDirective.decorators = [
    { type: Directive, args: [{
                selector: '[shContextMenuItem]'
            },] }
];
/** @nocollapse */
ShContextMenuItemDirective.ctorParameters = () => [
    { type: TemplateRef, decorators: [{ type: Optional }] }
];
ShContextMenuItemDirective.propDecorators = {
    subMenu: [{ type: Input }],
    divider: [{ type: Input }],
    visible: [{ type: Input }],
    disabled: [{ type: Input }],
    closeOnClick: [{ type: Input }],
    click: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    ShContextMenuItemDirective.prototype.subMenu;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.divider;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.visible;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.disabled;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.closeOnClick;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.click;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.context;
    /**
     * @type {?}
     * @private
     */
    ShContextMenuItemDirective.prototype._active;
    /** @type {?} */
    ShContextMenuItemDirective.prototype.template;
}
