// @gsp-lcm/metadatart-selector4biztype/esm5/metadata/metadata-rtservice4biztype.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-rtservice4biztype.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, LOCALE_ID } from '@angular/core';
import { GSPMetadataRTService4BizType } from './metadata.rtservice4biztype';
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { NotifyModule } from '@farris/ui-notify';
import { GspBoBizTypeMapping } from './gspbobiztypemapping';
import { GspMdBizTypeMapping } from './gspmdbiztypemapping';
import { MdBizTypeMappingDto } from './mdbiztypemappingdto';
import { GspMdBizTypeMappingType } from './gspmdbiztypemappingtype';
import { MetadataInfoComponent } from './metadatart-select4biztype/components/metadatainfo/metadatainfo.component';
import { CommonModule } from '@angular/common';
import { MetadataInfoList } from './metadatart-select4biztype/components/metadatainfolist/metadatainfolist.component';
import { DropDownsModule } from '@progress/kendo-angular-dropdowns';
import { FormsModule } from '@angular/forms';
import { LayoutModule } from '@progress/kendo-angular-layout';
import { BOTreePanelPackageModule } from '@gsp-lcm/bo-tree';
import { BORtServiceModule, BusinessObjectRtService } from '@gsp-lcm/bo-rt-service';
import { SplitterModule } from '@farris/ui-splitter';
import { FarrisSectionModule } from '@farris/ui-section';
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
import { TranslateResolveService } from './models/TranslateResolveService';
import { MetadataInfoService } from './metadatart-select4biztype/metadatainfo.service';
import { HttpService } from '@ecp-caf/caf-common';
import { DatagridModule } from '@farris/ui-datagrid';
import { FarrisFormsModule } from '@farris/ui-forms';
/**
 * @param {?} http
 * @return {?}
 */
export function createTranslateLoader(http) {
    return new TranslateHttpLoader(http, "/platform/runtime/common/web/@gsp-lcm/metadatart-selector4biztype/i18n/", ".json");
}
var ɵ0 = (createTranslateLoader), ɵ1 = localStorage.getItem('languageCode') || 'zh-CHS';
var GSPMetadataRTServiceModule4BizType = /** @class */ (function () {
    function GSPMetadataRTServiceModule4BizType() {
    }
    GSPMetadataRTServiceModule4BizType.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        FarrisFormsModule,
                        HttpClientModule,
                        NotifyModule.forRoot(),
                        CommonModule,
                        FormsModule,
                        LayoutModule,
                        DropDownsModule,
                        BOTreePanelPackageModule,
                        SplitterModule,
                        FarrisSectionModule,
                        BORtServiceModule,
                        DatagridModule,
                        TranslateModule.forRoot({
                            loader: {
                                provide: TranslateLoader,
                                useFactory: ɵ0,
                                deps: [HttpClient]
                            }
                        }),
                    ],
                    declarations: [MetadataInfoComponent, MetadataInfoList],
                    exports: [MetadataInfoComponent],
                    providers: [
                        HttpService,
                        GSPMetadataRTService4BizType,
                        MetadataInfoService,
                        BusinessObjectRtService,
                        TranslateResolveService,
                        GspBoBizTypeMapping,
                        GspMdBizTypeMapping,
                        MdBizTypeMappingDto,
                        GspMdBizTypeMappingType,
                        { provide: LOCALE_ID, useValue: ɵ1 }
                    ],
                    entryComponents: [MetadataInfoComponent]
                },] }
    ];
    return GSPMetadataRTServiceModule4BizType;
}());
export { GSPMetadataRTServiceModule4BizType };
export { ɵ0, ɵ1 };
