// @gsp-lcm/metadatart-selector4biztype/esm5/metadata/metadataprojectdto.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadataprojectdto.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MetadataProjectDto = /** @class */ (function () {
    // tslint:disable-next-line:max-line-length
    function MetadataProjectDto(packageName, projectPath, projectName, projectNameSpace, depPackageName, depPackageVersion, depPackageLocation) {
        this.packageName = packageName;
        this.projectName = projectName;
        this.projectPath = projectPath;
        this.projectNameSpace = projectNameSpace;
        this.depPackageName = depPackageName;
        this.depPackageVersion = depPackageVersion;
        this.depPackageLocation = depPackageLocation;
    }
    return MetadataProjectDto;
}());
export { MetadataProjectDto };
if (false) {
    /** @type {?} */
    MetadataProjectDto.prototype.packageName;
    /** @type {?} */
    MetadataProjectDto.prototype.projectPath;
    /** @type {?} */
    MetadataProjectDto.prototype.projectName;
    /** @type {?} */
    MetadataProjectDto.prototype.projectNameSpace;
    /** @type {?} */
    MetadataProjectDto.prototype.depPackageName;
    /** @type {?} */
    MetadataProjectDto.prototype.depPackageVersion;
    /** @type {?} */
    MetadataProjectDto.prototype.depPackageLocation;
}
