// @qdp/spread/esm5/lib/util/render.extends.utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ServicesUtils } from "./services.utils";
var RenderExtendsUtils = /** @class */ (function () {
    function RenderExtendsUtils(_queryId) {
        this.init(_queryId);
    }
    /**
     * @private
     * @param {?} _queryId
     * @return {?}
     */
    RenderExtendsUtils.prototype.init = /**
     * @private
     * @param {?} _queryId
     * @return {?}
     */
    function (_queryId) {
        this.dataRenderExtendService = ServicesUtils.getServices('dataRenderExtendService' + _queryId);
        this.renderExtends = ServicesUtils.getServices('renderExtends' + _queryId);
    };
    /**
   * 执行查询扩展
   */
    /**
     * 执行查询扩展
     * @param {?} extendCode
     * @param {?} extendName
     * @param {?} extendData
     * @param {?} control
     * @return {?}
     */
    RenderExtendsUtils.prototype.runRenderExtends = /**
     * 执行查询扩展
     * @param {?} extendCode
     * @param {?} extendName
     * @param {?} extendData
     * @param {?} control
     * @return {?}
     */
    function (extendCode, extendName, extendData, control) {
        /** @type {?} */
        var $event = this.createRenderExtendParams({
            eventCode: extendCode,
            // 'beforeQueryData',
            eventName: extendName,
            // this.localizeService.getValue('spread.message.extend.query.before'),
            data: extendData,
            // params,
            spread: control,
            // obj.spread
            control: control // obj.spread
        });
        if (extendCode === 'beforeQueryData') {
            if (this.dataRenderExtendService && this.dataRenderExtendService.beforeQueryData) {
                this.dataRenderExtendService.beforeQueryData($event);
            }
            else {
                if (this.renderExtends && this.renderExtends.beforeQueryData) {
                    this.renderExtends.beforeQueryData($event);
                }
            }
        }
        else if (extendCode === 'beforeLoadData') {
            if (this.dataRenderExtendService && this.dataRenderExtendService.beforeLoadData) {
                this.dataRenderExtendService.beforeLoadData($event);
            }
            else {
                if (this.renderExtends && this.renderExtends.beforeQueryData) {
                    this.renderExtends.beforeLoadData($event);
                }
            }
        }
        else if (extendCode === 'afterLoadData') {
            if (this.dataRenderExtendService && this.dataRenderExtendService.afterLoadData) {
                this.dataRenderExtendService.afterLoadData($event);
            }
            else {
                if (this.renderExtends && this.renderExtends.beforeQueryData) {
                    this.renderExtends.afterLoadData($event);
                }
            }
        }
        return $event['data'];
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RenderExtendsUtils.prototype.createRenderExtendParams = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        return {
            eventCode: obj.eventCode,
            eventName: obj.eventName,
            data: obj.data,
            control: obj.control,
            spread: obj.spread || undefined
        };
    };
    return RenderExtendsUtils;
}());
export { RenderExtendsUtils };
if (false) {
    /** @type {?} */
    RenderExtendsUtils.prototype.dataRenderExtendService;
    /** @type {?} */
    RenderExtendsUtils.prototype.renderExtends;
}
