// @qdp/spread/esm5/lib/util/schema.utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { RtfServices } from "@qdp/common";
import { of } from "rxjs";
import { switchMap } from "rxjs/operators";
var SchemaUtils = /** @class */ (function () {
    function SchemaUtils(_serviceCache) {
        this.init(_serviceCache);
    }
    /**
     * @private
     * @param {?} _serviceCache
     * @return {?}
     */
    SchemaUtils.prototype.init = /**
     * @private
     * @param {?} _serviceCache
     * @return {?}
     */
    function (_serviceCache) {
        this.schemaManager = _serviceCache.schemaManagerService;
        this.cacheService = _serviceCache.cacheService;
    };
    /**
     * 获取格式方案信息
     * @param option 参数配置项
     * { schemaId: 方案ID, schemaValue: 格式方案（交叉表時快速設置時使用）, controlType: 查询类型：list  treelist  crosstab, queryId: 查询ID, queryRelativeUrl：eapi基础uri, groupType: 方案分组标识 }
     * @returns 方案列表
     */
    /**
     * 获取格式方案信息
     * @param {?} option 参数配置项
     * { schemaId: 方案ID, schemaValue: 格式方案（交叉表時快速設置時使用）, controlType: 查询类型：list  treelist  crosstab, queryId: 查询ID, queryRelativeUrl：eapi基础uri, groupType: 方案分组标识 }
     * @return {?} 方案列表
     */
    SchemaUtils.prototype.getSchemaInfo = /**
     * 获取格式方案信息
     * @param {?} option 参数配置项
     * { schemaId: 方案ID, schemaValue: 格式方案（交叉表時快速設置時使用）, controlType: 查询类型：list  treelist  crosstab, queryId: 查询ID, queryRelativeUrl：eapi基础uri, groupType: 方案分组标识 }
     * @return {?} 方案列表
     */
    function (option) {
        /** @type {?} */
        var self = this;
        option.schemaId = option.schemaId === 'prefab' || option.schemaId === '1'
            ? '1'
            : option.schemaValue && option.schemaValue.id && option.controlType === 'crosstab'
                ? option.schemaValue.id
                : option.schemaId;
        if (!option.schemaId) {
            return self.getDefaultSchemaInfo(option);
        }
        else {
            /** @type {?} */
            var organizationId = this.cacheService.get(RtfServices.getTabId(option.queryId) + 'organizationId') || '';
            /** @type {?} */
            var result$ = this.schemaManager.getSchema(option.schemaId, option.queryId, option.queryRelativeUrl, option.groupType, organizationId, option.fromCache);
            return result$.pipe(switchMap((/**
             * @param {?} schemaValue
             * @return {?}
             */
            function (schemaValue) {
                if (schemaValue && schemaValue['id']) {
                    return of(schemaValue);
                }
            })));
        }
    };
    /**
     * 获取默认格式方案
     * @param option 参数配置项
     * { queryId：查询ID, queryRelativeUrl：eapi基础uri, groupType: 方案分组标识 }
     * @returns 默认格式方案
     */
    /**
     * 获取默认格式方案
     * @param {?} option 参数配置项
     * { queryId：查询ID, queryRelativeUrl：eapi基础uri, groupType: 方案分组标识 }
     * @return {?} 默认格式方案
     */
    SchemaUtils.prototype.getDefaultSchemaInfo = /**
     * 获取默认格式方案
     * @param {?} option 参数配置项
     * { queryId：查询ID, queryRelativeUrl：eapi基础uri, groupType: 方案分组标识 }
     * @return {?} 默认格式方案
     */
    function (option) {
        /** @type {?} */
        var organizationId = this.cacheService.get(RtfServices.getTabId(option.queryId) + 'organizationId') || '';
        return this.schemaManager.getDefaultSchema(option.queryId, option.queryRelativeUrl, option.groupType, organizationId);
    };
    return SchemaUtils;
}());
export { SchemaUtils };
if (false) {
    /** @type {?} */
    SchemaUtils.prototype.schemaManager;
    /** @type {?} */
    SchemaUtils.prototype.cacheService;
}
