// @farris/ui-datagrid/esm5/lib/plugins/smart-filter/controlData/filter-control.options.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} col
 * @return {?}
 */
export function TextFilterControl(col) {
    return {
        controltype: 'text'
    };
}
/**
 * @param {?} col
 * @return {?}
 */
export function BooleanFilterControl(col) {
    return {
        controltype: 'bool-check',
    };
}
/**
 * @param {?} col
 * @param {?} enumopts
 * @return {?}
 */
export function EnumFilterControl(col, enumopts) {
    /** @type {?} */
    var valueField = enumopts.valueField || enumopts.idField;
    /** @type {?} */
    var data = enumopts.data.map((/**
     * @param {?} n
     * @return {?}
     */
    function (n) {
        return {
            value: n[valueField],
            name: n[enumopts.textField]
        };
    }));
    return {
        controltype: 'checkboxgroup',
        enumValues: data,
    };
}
/**
 * @param {?} col
 * @param {?} options
 * @return {?}
 */
export function DateFilterControl(col, options) {
    /** @type {?} */
    var tye = 'datetime';
    /** @type {?} */
    var fmt = 'yyyy-MM-dd';
    /** @type {?} */
    var returnFmt = 'yyyy-MM-dd';
    if (options && options.format) {
        fmt = options.format;
        if (fmt.indexOf('HH:') == -1) {
            tye = 'date';
        }
        else {
            returnFmt += ' HH:mm';
            if (fmt.indexOf('ss') > -1) {
                returnFmt += ':ss';
            }
        }
    }
    else {
        tye = 'date';
    }
    // all = 1, 年月日
    // noDate = 2,  年月
    // noDateAndMonth = 3,  年
    return {
        controltype: 'flexibleDate',
        dateFormat: fmt,
        single: false,
        showType: 1,
        showTime: tye === 'datetime'
    };
}
/**
 * @param {?} col
 * @return {?}
 */
export function NumberFilterControl(col) {
    return {
        controltype: 'flexibleNumber',
        single: false
    };
}
