// @farris/ui-filter/esm5/lib/entity/conditionvalue/checkboxgroupvalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
var CheckboxGroupValue = /** @class */ (function () {
    function CheckboxGroupValue(value) {
        if (value === void 0) { value = { value: [] }; }
        this.clear$ = new EventEmitter();
        this.value = value.value;
    }
    /**
     * @return {?}
     */
    CheckboxGroupValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = [];
        this.clear$.emit();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    CheckboxGroupValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = value;
    };
    /**
     * @return {?}
     */
    CheckboxGroupValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.value;
    };
    /**
     * @return {?}
     */
    CheckboxGroupValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.CheckboxGroup;
    };
    /**
     * @return {?}
     */
    CheckboxGroupValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return this.value.length == 0;
    };
    /**
     * @return {?}
     */
    CheckboxGroupValue.prototype.getValueText = /**
     * @return {?}
     */
    function () {
        return this.value.join(',');
    };
    return CheckboxGroupValue;
}());
export { CheckboxGroupValue };
if (false) {
    /** @type {?} */
    CheckboxGroupValue.prototype.value;
    /** @type {?} */
    CheckboxGroupValue.prototype.clear$;
}
