// @farris/ui-filter/esm5/lib/entity/conditionvalue/flexiblenumbervalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
var FlexibleNumberValue = /** @class */ (function () {
    function FlexibleNumberValue(value, single) {
        if (value === void 0) { value = { startValue: null, endValue: null, value: null }; }
        if (single === void 0) { single = true; }
        this.clear$ = new EventEmitter();
        this.startValue = value.startValue;
        this.endValue = value.endValue;
        this.value = value.value;
        this.single = single;
    }
    /**
     * @return {?}
     */
    FlexibleNumberValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.FlexibleNumber;
    };
    /**
     * @return {?}
     */
    FlexibleNumberValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        if (this.single) {
            return !this.value;
        }
        else
            return (this.startValue == null && this.endValue == null);
    };
    /**
     * @return {?}
     */
    FlexibleNumberValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.startValue = undefined;
        this.endValue = undefined;
        this.value = undefined;
        this.clear$.emit();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    FlexibleNumberValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        throw new Error("Method not implemented.");
    };
    /**
     * @return {?}
     */
    FlexibleNumberValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return {
            begin: this.startValue,
            end: this.endValue,
            value: this.value
        };
    };
    /**
     * @return {?}
     */
    FlexibleNumberValue.prototype.getValueText = /**
     * @return {?}
     */
    function () {
        if (this.single) {
            if (this.value == undefined || (!this.value && typeof (this.value) != 'undefined' && this.value != 0)) {
                return this.value;
            }
            else {
                return '' + this.value;
            }
        }
        else {
            if (this.startValue == undefined && this.endValue == undefined) {
                return '';
            }
            else if (this.startValue == undefined) {
                return ('~' + this.endValue);
            }
            else if (this.endValue == undefined) {
                return (this.startValue + '~');
            }
            else {
                return (this.startValue + '~' + this.endValue);
            }
        }
    };
    return FlexibleNumberValue;
}());
export { FlexibleNumberValue };
if (false) {
    /** @type {?} */
    FlexibleNumberValue.prototype.startValue;
    /** @type {?} */
    FlexibleNumberValue.prototype.endValue;
    /** @type {?} */
    FlexibleNumberValue.prototype.value;
    /** @type {?} */
    FlexibleNumberValue.prototype.single;
    /** @type {?} */
    FlexibleNumberValue.prototype.clear$;
}
