// @farris/ui-filter/esm5/lib/entity/conditionvalue/helpvalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
import { cloneDeep } from 'lodash-es';
var HelpValue = /** @class */ (function () {
    function HelpValue(data) {
        if (data === void 0) { data = { value: [], valueField: undefined, textValue: undefined }; }
        this.clear$ = new EventEmitter();
        if (data.hasOwnProperty('textValue')) { // 新结构，value是数组
            this.valueField = data.valueField;
            this.value = data.value;
            this.textValue = data.textValue;
        }
        else { // 旧结构，value是对象
            this.valueField = data.valueField;
            this.value = data.value ? [data.value] : [];
            this.textValue = 'help-text-value-null'; // 旧结构无textValue属性
        }
    }
    /**
     * @param {?} helpItem
     * @param {?} args
     * @return {?}
     */
    HelpValue.getPropValue = /**
     * @param {?} helpItem
     * @param {?} args
     * @return {?}
     */
    function (helpItem, args) {
        if (args.length > 1) {
            /** @type {?} */
            var arg = args.shift();
            return helpItem[arg] ? this.getPropValue(helpItem[arg], args) : null;
        }
        return helpItem[args[0]];
    };
    /**
     * @return {?}
     */
    HelpValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = [];
        this.valueField = undefined;
        this.textValue = undefined;
        this.clear$.emit();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    HelpValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        throw new Error("Method not implemented.");
    };
    /**
     * @return {?}
     */
    HelpValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        if (this.valueField) {
            /** @type {?} */
            var args_1 = this.valueField.split('.');
            /** @type {?} */
            var valueArr = this.value.map((/**
             * @param {?} helpItem
             * @return {?}
             */
            function (helpItem) { return HelpValue.getPropValue(helpItem, cloneDeep(args_1)); }));
            return valueArr && valueArr.length ? valueArr.join(',') : '';
        }
        else {
            return '';
        }
    };
    /**
     * @return {?}
     */
    HelpValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.Help;
    };
    /**
     * @return {?}
     */
    HelpValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.valueField;
    };
    /**
     * @param {?} textField
     * @return {?}
     */
    HelpValue.prototype.getTextValue = /**
     * @param {?} textField
     * @return {?}
     */
    function (textField) {
        /** @type {?} */
        var args = textField.split('.');
        /** @type {?} */
        var textArr = this.value.map((/**
         * @param {?} helpItem
         * @return {?}
         */
        function (helpItem) { return HelpValue.getPropValue(helpItem, cloneDeep(args)); }));
        return textArr && textArr.length ? textArr.join(',') : '';
    };
    /**
     * @return {?}
     */
    HelpValue.prototype.getValueText = /**
     * @return {?}
     */
    function () {
        return this.textValue;
    };
    return HelpValue;
}());
export { HelpValue };
if (false) {
    /** @type {?} */
    HelpValue.prototype.valueField;
    /** @type {?} */
    HelpValue.prototype.value;
    /** @type {?} */
    HelpValue.prototype.textValue;
    /** @type {?} */
    HelpValue.prototype.clear$;
}
