// @farris/ui-filter/esm5/lib/entity/conditionvalue/ivalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
import { TextValue } from "./textvalue";
import { NumberRangeValueConvertor } from "./numberrangevaue";
import { DateRangeValue } from "./daterangevalue";
import { HelpValue } from "./helpvalue";
import { DropDownListValueConvert } from "./dropdownlistvalue";
/**
 * @record
 */
export function IValue() { }
if (false) {
    /**
     * @return {?}
     */
    IValue.prototype.getControlType = function () { };
    /**
     * @return {?}
     */
    IValue.prototype.getOriginalValue = function () { };
    /**
     * @param {?} value
     * @return {?}
     */
    IValue.prototype.setOriginalValue = function (value) { };
    /**
     * @return {?}
     */
    IValue.prototype.clearValue = function () { };
    /**
     * @return {?}
     */
    IValue.prototype.isEmpty = function () { };
    /**
     * @param {?=} ctrl
     * @return {?}
     */
    IValue.prototype.getValueText = function (ctrl) { };
}
var ConditionValueConvert = /** @class */ (function () {
    function ConditionValueConvert() {
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    ConditionValueConvert.prototype.convertToObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var conditionValue = (/** @type {?} */ (obj));
        /** @type {?} */
        var controlType = conditionValue.getControlType();
        /** @type {?} */
        var jobj = new Object();
        jobj["Type"] = controlType;
        if (controlType == ControlType.DropDown) {
            /** @type {?} */
            var convertor = new DropDownListValueConvert();
            jobj["Content"] = convertor.convertToObject(conditionValue);
        }
        else if (controlType == ControlType.NumberRange) {
            /** @type {?} */
            var convertor = new NumberRangeValueConvertor();
            jobj["Content"] = convertor.convertToObject(conditionValue);
        }
        else {
            jobj["Content"] = conditionValue;
        }
        return jobj;
    };
    /**
     * @param {?} jobj
     * @return {?}
     */
    ConditionValueConvert.prototype.initFromObject = /**
     * @param {?} jobj
     * @return {?}
     */
    function (jobj) {
        /** @type {?} */
        var controlType = (/** @type {?} */ (jobj["Type"]));
        /** @type {?} */
        var content = jobj["Content"];
        if (controlType == ControlType.Text) {
            return Object.assign(new TextValue(), (/** @type {?} */ (content)));
        }
        else if (controlType == ControlType.NumberRange) {
            /** @type {?} */
            var convertor = new NumberRangeValueConvertor();
            return convertor.initFromObject(content);
        }
        else if (controlType == ControlType.DateRange) {
            return Object.assign(new DateRangeValue(), (/** @type {?} */ (content)));
        }
        else if (controlType == ControlType.DropDown) {
            /** @type {?} */
            var convertor = new DropDownListValueConvert();
            return convertor.initFromObject(content);
        }
        else if (controlType == ControlType.Help) {
            return Object.assign(new HelpValue(), (/** @type {?} */ (content)));
        }
    };
    return ConditionValueConvert;
}());
export { ConditionValueConvert };
