// @farris/ui-filter/esm5/lib/entity/conditionvalue/numbervalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
var NumberValue = /** @class */ (function () {
    function NumberValue(value) {
        if (value === void 0) { value = { value: null }; }
        this.clear$ = new EventEmitter();
        this.value = value.value;
    }
    /**
     * @return {?}
     */
    NumberValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.Number;
    };
    /**
     * @return {?}
     */
    NumberValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = undefined;
        this.clear$.emit();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NumberValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        throw new Error("Method not implemented.");
    };
    /**
     * @return {?}
     */
    NumberValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.value;
    };
    /**
     * @return {?}
     */
    NumberValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.value;
    };
    /**
     * @return {?}
     */
    NumberValue.prototype.getValueText = /**
     * @return {?}
     */
    function () {
        if (this.value == undefined || (!this.value && typeof (this.value) != 'undefined' && this.value != 0)) {
            return this.value;
        }
        else {
            return '' + this.value;
        }
    };
    return NumberValue;
}());
export { NumberValue };
if (false) {
    /** @type {?} */
    NumberValue.prototype.value;
    /** @type {?} */
    NumberValue.prototype.clear$;
}
