// @farris/ui-filter/esm5/lib/entity/conditionvalue/searchvalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
var SearchValue = /** @class */ (function () {
    function SearchValue(value) {
        if (value === void 0) { value = { value: '' }; }
        this.clear$ = new EventEmitter();
        this.value = value.value;
    }
    /**
     * @return {?}
     */
    SearchValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.Search;
    };
    /**
     * @return {?}
     */
    SearchValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = undefined;
        this.clear$.emit();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    SearchValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        throw new Error("Method not implemented.");
    };
    /**
     * @return {?}
     */
    SearchValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.value;
    };
    /**
     * @return {?}
     */
    SearchValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.value;
    };
    /**
     * @return {?}
     */
    SearchValue.prototype.getValueText = /**
     * @return {?}
     */
    function () {
        return this.value;
    };
    return SearchValue;
}());
export { SearchValue };
if (false) {
    /** @type {?} */
    SearchValue.prototype.value;
    /** @type {?} */
    SearchValue.prototype.clear$;
}
