// @farris/ui-common/esm5/column/data-column.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-05-23 10:27:07
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-08-15 10:15:43
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { NgModule } from '@angular/core';
import { DateHelperModule } from '@farris/ui-common/date';
import { NumberHelperModule } from '@farris/ui-common/number';
import { ColumnFormatService } from './column-format.service';
import { ColumnFormatterPipe } from './data-formatter.pipe';
var DataColumnModule = /** @class */ (function () {
    function DataColumnModule() {
    }
    /**
     * @return {?}
     */
    DataColumnModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: DataColumnModule,
            providers: [
                ColumnFormatService
            ]
        };
    };
    DataColumnModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        DateHelperModule.forRoot(),
                        NumberHelperModule.forRoot()
                    ],
                    exports: [ColumnFormatterPipe],
                    declarations: [
                        ColumnFormatterPipe
                    ],
                    providers: [],
                },] }
    ];
    return DataColumnModule;
}());
export { DataColumnModule };
