// @farris/ui-tabs/esm5/lib/tooltip/tooltip.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NgbTooltip, NgbTooltipWindow } from './tooltip';
import { NgbTooltipConfig } from './tooltip-config';
export { NgbTooltipConfig } from './tooltip-config';
export { NgbTooltip } from './tooltip';
var NgbTooltipModule = /** @class */ (function () {
    function NgbTooltipModule() {
    }
    /**
     * @return {?}
     */
    NgbTooltipModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: NgbTooltipModule,
            providers: [NgbTooltipConfig]
        };
    };
    NgbTooltipModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [NgbTooltip, NgbTooltipWindow],
                    exports: [NgbTooltip],
                    entryComponents: [NgbTooltipWindow]
                },] }
    ];
    return NgbTooltipModule;
}());
export { NgbTooltipModule };
