// @farris/lookup-config/esm2015/lib/config.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/config.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class ConfigService {
    // httpUrl = 'http://10.25.11.67:5200';
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.getHelpConfig_Path = '/api/runtime/bcc/v1.0/helpconfig';
        this.saveHelpConfig_Path = '/api/runtime/bcc/v1.0/helpconfig';
    }
    /**
     * 通过configId获取回显数据
     * @param {?} configId 配置id
     * @return {?}
     */
    getHelpConfig(configId) {
        /** @type {?} */
        const url = this.getHelpConfig_Path + '/' + configId;
        return this.http.get(url);
        //return this.http.get('assets/data/helpconfig.json');
    }
    /**
     * 保存帮助配置
     * @param {?} params 参数
     * @return {?}
     */
    saveHelpConfig(params) {
        /** @type {?} */
        const saveUrl = this.saveHelpConfig_Path;
        /** @type {?} */
        const thisParams = Object.assign({}, params);
        thisParams.queryParam = JSON.stringify(thisParams.queryParam);
        //console.log('save',thisParams)
        return this.http.put(saveUrl, thisParams);
        //return this.http.get('assets/data/savehelpconfig.json');
    }
}
ConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ConfigService.ctorParameters = () => [
    { type: HttpClient }
];
/** @nocollapse */ ConfigService.ngInjectableDef = i0.defineInjectable({ factory: function ConfigService_Factory() { return new ConfigService(i0.inject(i1.HttpClient)); }, token: ConfigService, providedIn: "root" });
if (false) {
    /** @type {?} */
    ConfigService.prototype.getHelpConfig_Path;
    /** @type {?} */
    ConfigService.prototype.saveHelpConfig_Path;
    /**
     * @type {?}
     * @private
     */
    ConfigService.prototype.http;
}
