// @farris/lookup-config/esm2015/lib/lookup-data.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/lookup-data.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { NotifyService } from '@farris/ui-notify';
import { LoadingService } from '@farris/ui-loading';
export class LookupDataService {
    /**
     * @param {?} http
     * @param {?} injector
     * @param {?} notifyService
     * @param {?} loading
     */
    constructor(http, injector, notifyService, loading) {
        this.http = http;
        this.injector = injector;
        this.notifyService = notifyService;
        this.loading = loading;
    }
    /**
     * @param {?} url
     * @param {?} params
     * @return {?}
     */
    getData(url, params) {
        if (!params || !params.helpId) {
            this.notifyService.error('helpId不能为空');
            this.loading.clearAll();
            return;
        }
        /** @type {?} */
        const _url = `/api/runtime/bcc/v1.0/helpconfig/data/${params.helpId}`;
        params = Object.assign(params || {}, { withSpecificField: true });
        return this.http.get(_url, { params: { queryParam: JSON.stringify(params) } });
    }
}
LookupDataService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
LookupDataService.ctorParameters = () => [
    { type: HttpClient },
    { type: Injector },
    { type: NotifyService },
    { type: LoadingService }
];
if (false) {
    /** @type {?} */
    LookupDataService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    LookupDataService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    LookupDataService.prototype.notifyService;
    /**
     * @type {?}
     * @private
     */
    LookupDataService.prototype.loading;
}
