// @farris/devkit/esm2015/lib/misc/guid.js
// tslint:disable: max-line-length
import { UID } from './uid';
/**
 * GUID创建服务
 * @scope 静态类没有提供Provider
 */
export class Guid {
    constructor(guid) {
        if (!guid) {
            throw new TypeError('Invalid argument; `value` has no value.');
        }
        this.value = Guid.EMPTY;
        if (guid) {
            this.value = guid;
        }
    }
    static isGuid(guid) {
        const value = guid.toString();
        return guid && (guid instanceof Guid || Guid.validator.test(value));
    }
    static create() {
        return new Guid(UID.create());
    }
    static createEmpty() {
        return new Guid('');
    }
    static parse(guid) {
        return new Guid(guid);
    }
    static raw() {
        return UID.create();
    }
    equals(other) {
        // Comparing string `value` against provided `guid` will auto-call
        // toString on `guid` for comparison
        return Guid.isGuid(other) && this.value === other.toString();
    }
    isEmpty() {
        return this.value === Guid.EMPTY;
    }
    toString() {
        return this.value;
    }
    toJSON() {
        return {
            value: this.value,
        };
    }
}
Guid.validator = new RegExp('^[a-z0-9]+$', 'i');
Guid.EMPTY = '';
