// @farris/devkit/esm2015/lib/misc/router-param.service.js
import { Injectable } from '@angular/core';
/**
 * 路由参数服务
 */
class RouterParamService {
    /**
     * 获取参数
     */
    getParams(url) {
        const allParams = this.getAllParams();
        return allParams[url] || {};
    }
    /**
     * 设置参数
     */
    setParams(url, params) {
        const allParams = this.getAllParams();
        allParams[url] = params;
        this.setAllParams(allParams);
    }
    /**
     * 清空参数
     */
    clearParams() {
        throw new Error('Not Implemented');
    }
    /**
     * 获取全部参数
     */
    getAllParams() {
        const allParamsJson = window.sessionStorage.getItem('ROUTER_PARAMS') || '{}';
        const allParams = JSON.parse(allParamsJson);
        return allParams;
    }
    /**
     * 设置所有的参数
     */
    setAllParams(allParams) {
        allParams = allParams || {};
        const allParamsJson = JSON.stringify(allParams);
        window.sessionStorage.setItem('ROUTER_PARAMS', allParamsJson);
    }
    /**
     * 清空所有参数
     */
    clearAllParams() {
        window.sessionStorage.setItem('ROUTER_PARAMS', '{}');
    }
}
RouterParamService.decorators = [
    { type: Injectable }
];
export { RouterParamService };
