// @gsp-bef/gsp-be-metadata/esm5/lib/biz-entity/src/Util/BeOperationUtil.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
var BeOperationUtil = /** @class */ (function () {
    function BeOperationUtil() {
    }
    /**
     * @param {?} be
     * @return {?}
     */
    BeOperationUtil.prototype.getAllMgrActions = /**
     * @param {?} be
     * @return {?}
     */
    function (be) {
        return be.BizMgrActions;
    };
    /**
     * bizActions
     */
    /**
     * bizActions
     * @param {?} be
     * @return {?}
     */
    BeOperationUtil.prototype.getAllBizEntityActions = /**
     * bizActions
     * @param {?} be
     * @return {?}
     */
    function (be) {
        return this.getChidrenObjectsActions(be.MainObject);
    };
    /**
     * @param {?} bizObject
     * @return {?}
     */
    BeOperationUtil.prototype.getChidrenObjectsActions = /**
     * @param {?} bizObject
     * @return {?}
     */
    function (bizObject) {
        /** @type {?} */
        var result = new Array();
        if (!(bizObject)) {
            return result;
        }
        if (bizObject.BizActions && bizObject.BizActions.length > 0) {
            result.push.apply(result, tslib_1.__spread(bizObject.BizActions));
        }
        if (bizObject.ContainChildObjects && bizObject.ContainChildObjects.length > 0) {
            for (var i = 0; i < bizObject.ContainChildObjects.length; i++) {
                result.push.apply(result, tslib_1.__spread(this.getChidrenObjectsActions((/** @type {?} */ (bizObject.ContainChildObjects[i])))));
            }
        }
        return result;
    };
    /**
     * determination
     */
    /**
     * determination
     * @param {?} be
     * @return {?}
     */
    BeOperationUtil.prototype.getAllDeterminations = /**
     * determination
     * @param {?} be
     * @return {?}
     */
    function (be) {
        return this.getChidrenObjectsDeterminations(be.MainObject);
    };
    /**
     * @param {?} bizObject
     * @return {?}
     */
    BeOperationUtil.prototype.getChidrenObjectsDeterminations = /**
     * @param {?} bizObject
     * @return {?}
     */
    function (bizObject) {
        /** @type {?} */
        var result = new Array();
        if (!(bizObject)) {
            return result;
        }
        if (bizObject.Determinations && bizObject.Determinations.length > 0) {
            result.push.apply(result, tslib_1.__spread(bizObject.Determinations));
        }
        if (bizObject.ContainChildObjects && bizObject.ContainChildObjects.length > 0) {
            for (var i = 0; i < bizObject.ContainChildObjects.length; i++) {
                result.push.apply(result, tslib_1.__spread(this.getChidrenObjectsDeterminations((/** @type {?} */ (bizObject.ContainChildObjects[i])))));
            }
        }
        return result;
    };
    /**
     * validation
     */
    /**
     * validation
     * @param {?} be
     * @return {?}
     */
    BeOperationUtil.prototype.getAllValidations = /**
     * validation
     * @param {?} be
     * @return {?}
     */
    function (be) {
        return this.getChidrenObjectsValidations(be.MainObject);
    };
    /**
     * @param {?} bizObject
     * @return {?}
     */
    BeOperationUtil.prototype.getChidrenObjectsValidations = /**
     * @param {?} bizObject
     * @return {?}
     */
    function (bizObject) {
        /** @type {?} */
        var result = new Array();
        if (!(bizObject)) {
            return result;
        }
        if (bizObject.Validations && bizObject.Validations.length > 0) {
            result.push.apply(result, tslib_1.__spread(bizObject.Validations));
        }
        if (bizObject.ContainChildObjects && bizObject.ContainChildObjects.length > 0) {
            for (var i = 0; i < bizObject.ContainChildObjects.length; i++) {
                result.push.apply(result, tslib_1.__spread(this.getChidrenObjectsValidations((/** @type {?} */ (bizObject.ContainChildObjects[i])))));
            }
        }
        return result;
    };
    /**
     * 获取所有操作
     */
    /**
     * 获取所有操作
     * @param {?} be
     * @return {?}
     */
    BeOperationUtil.prototype.getAllOperations = /**
     * 获取所有操作
     * @param {?} be
     * @return {?}
     */
    function (be) {
        /** @type {?} */
        var result = new Array();
        result.push.apply(result, tslib_1.__spread(this.getAllBizEntityActions(be)));
        result.push.apply(result, tslib_1.__spread(this.getAllMgrActions(be)));
        result.push.apply(result, tslib_1.__spread(this.getAllDeterminations(be)));
        result.push.apply(result, tslib_1.__spread(this.getAllValidations(be)));
        return result;
    };
    /**
     * validation触发字段中是否包含某字段
     * @param eleId 字段ID
     */
    /**
     * validation触发字段中是否包含某字段
     * @param {?} be
     * @param {?} eleId 字段ID
     * @return {?}
     */
    BeOperationUtil.prototype.countainValidationRequestElement = /**
     * validation触发字段中是否包含某字段
     * @param {?} be
     * @param {?} eleId 字段ID
     * @return {?}
     */
    function (be, eleId) {
        /** @type {?} */
        var validations = this.getAllValidations(be);
        if (!!(validations) && validations.length > 0) {
            for (var i = 0; i < validations.length; i++) {
                /** @type {?} */
                var b = this.validationCountainRequestElement(validations[i], eleId);
                if (b) {
                    /** @type {?} */
                    var result = validations[i].Name;
                    return result;
                }
            }
        }
        return null;
    };
    /**
     * determination触发字段中是否包含某字段
     * @param eleId 字段ID
     */
    /**
     * determination触发字段中是否包含某字段
     * @param {?} be
     * @param {?} eleId 字段ID
     * @return {?}
     */
    BeOperationUtil.prototype.countainDeterminationRequestElement = /**
     * determination触发字段中是否包含某字段
     * @param {?} be
     * @param {?} eleId 字段ID
     * @return {?}
     */
    function (be, eleId) {
        /** @type {?} */
        var dtms = this.getAllDeterminations(be);
        if (!!(dtms) && dtms.length > 0) {
            for (var i = 0; i < dtms.length; i++) {
                /** @type {?} */
                var b = this.determinationCountainRequestElement(dtms[i], eleId);
                if (b) {
                    /** @type {?} */
                    var result = dtms[i].Name;
                    return result;
                }
            }
        }
        return null;
    };
    /**
     * @private
     * @param {?} id
     * @param {?} requestEles
     * @return {?}
     */
    BeOperationUtil.prototype.countainEle = /**
     * @private
     * @param {?} id
     * @param {?} requestEles
     * @return {?}
     */
    function (id, requestEles) {
        if (!!(id) && !!(requestEles) && requestEles.length > 0) {
            return requestEles.indexOf(id) >= 0;
        }
        return false;
    };
    /**
     * 触发字段中是否包含某字段
     */
    /**
     * 触发字段中是否包含某字段
     * @param {?} dtm
     * @param {?} id
     * @return {?}
     */
    BeOperationUtil.prototype.determinationCountainRequestElement = /**
     * 触发字段中是否包含某字段
     * @param {?} dtm
     * @param {?} id
     * @return {?}
     */
    function (dtm, id) {
        if (this.countainEle(id, dtm.RequestElements)) {
            return true;
        }
        if (!!(dtm.RequestChildElements) && dtm.RequestChildElements.length > 0) {
            return this.countainEle(id, dtm.RequestChildElements[0].RequestChildElementValue);
        }
        return false;
    };
    /**
     * 触发字段中是否包含某字段
     */
    /**
     * 触发字段中是否包含某字段
     * @param {?} validation
     * @param {?} id
     * @return {?}
     */
    BeOperationUtil.prototype.validationCountainRequestElement = /**
     * 触发字段中是否包含某字段
     * @param {?} validation
     * @param {?} id
     * @return {?}
     */
    function (validation, id) {
        if (this.countainEle(id, validation.RequestElements)) {
            return true;
        }
        if (!!(validation.RequestChildElements) && validation.RequestChildElements.length > 0) {
            return this.countainEle(id, validation.RequestChildElements[0].RequestChildElementValue);
        }
        return false;
    };
    return BeOperationUtil;
}());
export { BeOperationUtil };
