// @farris/devkit/esm5/lib/entity/validator/validation_types.js
var validationMessage = {};
/**
 * 验证类型定义
 */
var ValidationTypes = /** @class */ (function () {
    function ValidationTypes() {
    }
    /**
     * 判断验证类型是否合法
     * @param type 验证类型
     */
    ValidationTypes.isValidType = function (type) {
        var _this = this;
        return (type !== "isValidType" &&
            type !== 'getMessage' &&
            Object.keys(this)
                .map(function (key) { return _this[key]; })
                .indexOf(type) !== -1);
    };
    /**
     * 获取指定验证类型的提示信息
     * @param type 验证类型
     * @returns 返回验证信息
     */
    ValidationTypes.getMessage = function (type) {
        var messages = validationMessage[this.CURRENT_LANGUAGE] || validationMessage['zh-CHS'];
        return messages[type] || '';
    };
    ValidationTypes.setCurrentLanguage = function (language) {
        this.CURRENT_LANGUAGE = language;
    };
    ValidationTypes.CURRENT_LANGUAGE = 'zh-CNS';
    /**
     * 条件验证表达式
     */
    ValidationTypes.CUSTOM_VALIDATION = "customValidation";
    /**
     * 必填
     */
    ValidationTypes.REQUIRED = 'required';
    /**
     * 比较值是否相同
     */
    ValidationTypes.EQUALS = 'equals';
    /**
     * 比较值是否不相同
     */
    ValidationTypes.NOT_EQUALS = 'notEquals';
    /**
     * 数字
     */
    ValidationTypes.IS_NUMBER = 'isNumber';
    /**
     * 整数
     */
    ValidationTypes.IS_INT = 'isInt';
    /**
     * 浮点数
     */
    ValidationTypes.IS_FLOAT = 'isFloat';
    /**
     * 字符串
     */
    ValidationTypes.IS_STRING = 'isString';
    /**
     * 布尔值
     */
    ValidationTypes.IS_BOOLEAN = 'isBoolean';
    /**
     * 日期
     */
    ValidationTypes.IS_DATE = 'isDate';
    /**
     * 日期字符串
     */
    ValidationTypes.IS_DATE_STRING = 'isDateString';
    /**
     * 布尔值字符串
     */
    ValidationTypes.IS_BOOLEAN_STRING = 'isBooleanString';
    /**
     * 数字字符串
     */
    ValidationTypes.IS_NUMBER_STRING = 'isNumberString';
    /**
     * 邮箱地址
     */
    ValidationTypes.IS_EMAIL = 'isEmail';
    /**
     * JSON
     */
    ValidationTypes.IS_JSON = 'isJSON';
    /**
     * 字符串小写
     */
    ValidationTypes.IS_LOWERCASE = 'isLowercase';
    /**
     * 字符串大写
     */
    ValidationTypes.IS_UPPERCASE = 'isUppercase';
    /**
     * 数字范围
     */
    ValidationTypes.RANGE = 'range';
    /**
     * 最小值
     */
    ValidationTypes.MIN = 'min';
    /**
     * 最小值
     */
    ValidationTypes.MINVALUE = 'minValue';
    /**
     * 最大值
     */
    ValidationTypes.MAX = 'max';
    /**
     * 最大值
     */
    ValidationTypes.MAXVALUE = 'maxValue';
    /**
     * 长度
     */
    ValidationTypes.LENGTH = 'length';
    /**
     * 字符串最大长度
     */
    ValidationTypes.MAX_LENGTH = 'maxLength';
    /**
     * 字符串最小长度
     */
    ValidationTypes.MIN_LENGTH = 'minLength';
    /**
     * 最小日期
     */
    ValidationTypes.MIN_DATE = 'minDate';
    /**
     * 最大日期
     */
    ValidationTypes.MAX_DATE = 'maxDate';
    /**
     * 排除非法字符
     */
    ValidationTypes.EXCLUDE = 'exclude';
    /**
     * 正则表达式
     */
    ValidationTypes.MATCHES = 'matches';
    ValidationTypes.FIELD_CONTAINER = 'fieldContainer';
    return ValidationTypes;
}());
export { ValidationTypes };
validationMessage['zh-CHS'] = {
    fieldContainer: "$target \u7B2C $value \u884C",
    required: "\u8BF7\u8F93\u5165'$property'",
    equals: "'$property'\u7684\u503C\u4E0E$constraint1\u4E0D\u76F8\u7B49",
    notEquals: "'$property'\u7684\u503C\u4E0D\u80FD\u4E0E'$constraint1'\u76F8\u540C",
    isNumber: "'$property'\u7684\u503C\u4E0D\u662F\u6570\u5B57",
    isInt: "'$property'\u7684\u503C\u4E0D\u662F\u6574\u6570",
    isFloat: "'$property'\u7684\u503C\u4E0D\u662F\u6D6E\u70B9\u578B\u6570\u5B57",
    isBoolean: "'$property'\u7684\u503C\u4E0D\u662F\u5E03\u5C14\u503C",
    isDate: "'$property'\u7684\u503C\u4E0D\u662F\u6709\u6548\u65E5\u671F",
    isEmail: "\u90AE\u7BB1\u5730\u5740\u4E0D\u6B63\u786E",
    min: "'$property'\u7684\u503C\u4E0D\u5E94\u5C0F\u4E8E$constraint1",
    minValue: "'$property'\u7684\u503C\u4E0D\u5E94\u5C0F\u4E8E$constraint1",
    minDate: "'$property'\u7684\u65E5\u671F\u4E0D\u5E94\u65E9\u4E8E$constraint1",
    max: "'$property'\u7684\u503C\u4E0D\u5E94\u5927\u4E8E$constraint1",
    maxValue: "'$property'\u7684\u503C\u4E0D\u5E94\u5927\u4E8E$constraint1",
    maxDate: "'$property'\u4E0D\u5E94\u665A\u4E8E$constraint1",
    isBooleanString: "'$property'\u7684\u503C\u4E0D\u662F\u6709\u6548\u5E03\u5C14\u503C",
    isDateString: "'$property'\u7684\u503C\u4E0D\u662F\u6709\u6548\u7684\u65E5\u671F",
    isLowercase: "'$property'\u7684\u503C\u5E94\u5168\u90E8\u4E3A\u5C0F\u5199\u5B57\u7B26\u4E32",
    isUppercase: "'$property'\u7684\u503C\u5E94\u5168\u90E8\u4E3A\u5927\u5199\u5B57\u7B26\u4E32",
    length: "'$property'\u7684\u957F\u5EA6\u5E94\u4ECB\u4E8E$constraint1~$constraint2\u4E4B\u95F4",
    range: "'$property'\u7684\u503C\u5E94\u4ECB\u4E8E$constraint1~$constraint2\u4E4B\u95F4",
    maxLength: "'$property'\u7684\u957F\u5EA6\u4E0D\u5F97\u5927\u4E8E$constraint1",
    minLength: "'$property'\u7684\u957F\u5EA6\u4E0D\u5F97\u5C0F\u4E8E$constraint1",
    isNumberString: "'$property'\u7684\u503C\u4E0D\u662F\u6570\u5B57",
    exclude: "'$property'\u7684\u503C\u4E0D\u80FD\u5305\u542B\uFF1A$constraint1",
    matches: "'$property'\u6821\u9A8C\u4E0D\u901A\u8FC7"
};
validationMessage['en'] = {
    fieldContainer: "$target row $value",
    required: "Please input '$property'",
    equals: "'$property' should equals '$constraint1'",
    notEquals: "'$property' should not equals '$constraint1'",
    isNumber: "'$property' should be number",
    isInt: "'$property' should be integer",
    isFloat: "'$property' should be float",
    isBoolean: "'$property' should be boolean",
    isDate: "'$property' should be date",
    isEmail: "'$property' should be e-mail address",
    min: "'$property' should not less than $constraint1",
    minValue: "'$property' should not less than $constraint1",
    minDate: "'$property' should not early than $constraint1",
    max: "'$property' should not bigger than $constraint1",
    maxValue: "'$property' should not bigger than $constraint1",
    maxDate: "'$property' should not late than $constraint1",
    isBooleanString: "'$property' should be boolean string",
    isDateString: "'$property' should be date string",
    isLowercase: "'$property' should be lowercase charactor",
    isUppercase: "'$property' should be uppercase charactor",
    length: "'$property' length should between $constraint1~$constraint2\u4E4B\u95F4",
    range: "'$property' value should between $constraint1~$constraint2\u4E4B\u95F4",
    maxLength: "'$property' should not longer than $constraint1",
    minLength: "'$property' should not shorter than $constraint1",
    isNumberString: "'$property' should be number string",
    exclude: "'$property' should not include $constraint1",
    matches: "'$property' calibration failed"
};
validationMessage['zh-CHT'] = {
    fieldContainer: "$target \u7B2C $value \u884C",
    required: "\u8ACB\u8F38\u5165'$property'",
    equals: "'$property'\u7684\u503C\u8207$constraint1\u4E0D\u76F8\u7B49",
    notEquals: "'$property'\u7684\u503C\u4E0D\u80FD\u8207'$constraint1'\u76F8\u540C",
    isNumber: "'$property'\u7684\u503C\u4E0D\u662F\u6578\u5B57",
    isInt: "'$property'\u7684\u503C\u4E0D\u662F\u6574\u6578",
    isFloat: "'$property'\u7684\u503C\u4E0D\u662F\u6D6E\u9EDE\u578B\u6578\u5B57",
    isBoolean: "'$property'\u7684\u503C\u4E0D\u662F\u4F48\u723E\u503C",
    isDate: "'$property'\u7684\u503C\u4E0D\u662F\u6709\u6548\u65E5\u671F",
    isEmail: "\u90F5\u7BB1\u5730\u5740\u4E0D\u6B63\u78BA",
    min: "'$property'\u7684\u503C\u4E0D\u61C9\u5C0F\u65BC$constraint1",
    minValue: "'$property'\u7684\u503C\u4E0D\u61C9\u5C0F\u65BC$constraint1",
    minDate: "'$property'\u7684\u65E5\u671F\u4E0D\u61C9\u65E9\u65BC$constraint1",
    max: "'$property'\u7684\u503C\u4E0D\u61C9\u5927\u65BC$constraint1",
    maxValue: "'$property'\u7684\u503C\u4E0D\u61C9\u5927\u65BC$constraint1",
    maxDate: "'$property'\u4E0D\u61C9\u665A\u65BC$constraint1",
    isBooleanString: "'$property'\u7684\u503C\u4E0D\u662F\u6709\u6548\u4F48\u723E\u503C",
    isDateString: "'$property'\u7684\u503C\u4E0D\u662F\u6709\u6548\u7684\u65E5\u671F",
    isLowercase: "'$property'\u7684\u503C\u61C9\u5168\u90E8\u7232\u5C0F\u51A9\u5B57\u7B26\u4E32",
    isUppercase: "'$property'\u7684\u503C\u61C9\u5168\u90E8\u7232\u5927\u51A9\u5B57\u7B26\u4E32",
    length: "'$property'\u7684\u9577\u5EA6\u61C9\u4ECB\u65BC$constraint1~$constraint2\u4E4B\u9593",
    range: "'$property'\u7684\u503C\u61C9\u4ECB\u65BC$constraint1~$constraint2\u4E4B\u9593",
    maxLength: "'$property'\u7684\u9577\u5EA6\u4E0D\u5F97\u5927\u65BC$constraint1",
    minLength: "'$property'\u7684\u9577\u5EA6\u4E0D\u5F97\u5C0F\u65BC$constraint1",
    isNumberString: "'$property'\u7684\u503C\u4E0D\u662F\u6578\u5B57",
    exclude: "'$property'\u7684\u503C\u4E0D\u80FD\u5305\u542B\uFF1A$constraint1",
    matches: "'$property'\u6821\u9A57\u4E0D\u901A\u904E"
};
