// @gsp-wf/wf-process-editor/esm5/lib/views/bill-type-help/entity/DocuTypeHelperEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
var DocuTypeHelperEntity = /** @class */ (function (_super) {
    tslib_1.__extends(DocuTypeHelperEntity, _super);
    function DocuTypeHelperEntity() {
        return _super.call(this) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    DocuTypeHelperEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.id = null;
        this.moduleName = null;
        this.bizKindName = null;
        this.bizTypeName = null;
        this.code = null;
        this.name = null;
    };
    /**
     * @return {?}
     */
    DocuTypeHelperEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "id", this.id);
        BpmnModelJsonHelper.SetAttributeValue(obj, "moduleName", this.moduleName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizKindName", this.bizKindName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeName", this.bizTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    DocuTypeHelperEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.id = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "id");
        this.moduleName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "moduleName");
        this.bizKindName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizKindName");
        this.bizTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeName");
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
    };
    return DocuTypeHelperEntity;
}(JsonConvertableImpl));
export { DocuTypeHelperEntity };
if (false) {
    /** @type {?} */
    DocuTypeHelperEntity.prototype.id;
    /** @type {?} */
    DocuTypeHelperEntity.prototype.moduleName;
    /** @type {?} */
    DocuTypeHelperEntity.prototype.bizKindName;
    /** @type {?} */
    DocuTypeHelperEntity.prototype.bizTypeName;
    /** @type {?} */
    DocuTypeHelperEntity.prototype.code;
    /** @type {?} */
    DocuTypeHelperEntity.prototype.name;
}
