// @farris/command-services/esm5/lib/work-flow/work-flow-message.service.js
import { Injectable } from "@angular/core";
var WorkFlowMessageService = /** @class */ (function () {
    function WorkFlowMessageService() {
        this.messageHandler = function (callback) {
            return function (event) {
                var message = event.data;
                if (message) {
                    message.sender = event.source;
                }
                if (typeof callback === 'function') {
                    callback(message);
                }
            };
        };
    }
    /**
     * 监听工作流消息
     * @param callback 回调
     */
    WorkFlowMessageService.prototype.listen = function (callback) {
        this.handler = this.messageHandler(callback);
        window.addEventListener('message', this.handler);
    };
    WorkFlowMessageService.prototype.destroy = function () {
        if (this.handler) {
            window.removeEventListener('message', this.handler);
        }
    };
    /**
     * 给工作流表单发送消息
     * @param message
     */
    WorkFlowMessageService.prototype.send = function (message) {
        if (!message) {
            return;
        }
        var win = window;
        if (message.target) {
            win = message.target || window;
            delete message.target;
        }
        if (win) {
            win.postMessage(message, '*');
        }
    };
    WorkFlowMessageService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    WorkFlowMessageService.ctorParameters = function () { return []; };
    return WorkFlowMessageService;
}());
export { WorkFlowMessageService };
