// @farris/command-services/esm5/lib/work-flow/work-flow-message.js
import { Injectable, Injector } from "@angular/core";
import { WorkFlowMessageService } from "./work-flow-message.service";
/**
 * 工作流消息
 * @warning 禁止使用表单上下文，所有上下文必须从外部传入
 */
var WorkFlowMessage = /** @class */ (function () {
    function WorkFlowMessage(injector, workFlowMessageService) {
        this.injector = injector;
        this.workFlowMessageService = workFlowMessageService;
        this.listeners = new Map();
    }
    WorkFlowMessage.prototype.setup = function () {
        var _this = this;
        if (!this.workFlowMessageService) {
            return;
        }
        this.workFlowMessageService.listen(function (message) {
            if (!_this.listeners || _this.listeners.size < 1) {
                return;
            }
            _this.listeners.forEach(function (callback, key) {
                if (typeof callback === 'function') {
                    callback(message);
                }
            });
        });
    };
    WorkFlowMessage.prototype.addEventListener = function (callback) {
        var id = new Date().valueOf().toString(16) + "-" + Math.ceil(Math.random() * 100);
        this.listeners.set(id, callback);
        return id;
    };
    WorkFlowMessage.prototype.removeEventListener = function (eventId) {
        if (!this.listeners || this.listeners.size < 1) {
            return;
        }
        this.listeners.delete(eventId);
        eventId = null;
    };
    WorkFlowMessage.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    WorkFlowMessage.ctorParameters = function () { return [
        { type: Injector },
        { type: WorkFlowMessageService }
    ]; };
    return WorkFlowMessage;
}());
export { WorkFlowMessage };
