// @gsp-svc/file-viewer/esm5/lib/gsp-document/gsp-document.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { GspDocumentComponent } from './gsp-document/gsp-document.component';
import { GspPageComponent } from './gsp-page/gsp-page.component';
import { ZoomService } from '../service/zoom.service';
import { ZoomDirective } from '../service/zoom.directive';
import { SelectionService } from '../service/selection.service';
import { SearchService } from '../service/search.service';
import { SearchableDirective } from '../service/searchable.directive';
import { RotationDirective } from '../service/rotation.directive';
import { SanitizeHtmlPipe, SanitizeResourceHtmlPipe, SanitizeStylePipe } from '../service/pipes';
import { PagePreloadService } from '../service/page-preload.service';
import { ScrollableDirective } from '../service/scrollable.directive';
import { ExcelDocumentComponent } from '../excel-document/excel-document.component';
import { ExcelPageComponent } from '../excel-page/excel-page.component';
var GspDocumentModule = /** @class */ (function () {
    function GspDocumentModule() {
    }
    GspDocumentModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [GspDocumentComponent, GspPageComponent, ExcelDocumentComponent, ExcelPageComponent,
                        ZoomDirective, SearchableDirective,
                        RotationDirective, SanitizeHtmlPipe, SanitizeResourceHtmlPipe, SanitizeStylePipe, ScrollableDirective],
                    imports: [
                        CommonModule
                    ],
                    exports: [GspDocumentComponent, ExcelDocumentComponent, ZoomDirective, SearchableDirective],
                    entryComponents: [GspDocumentComponent, ExcelDocumentComponent],
                    providers: [
                        ZoomService,
                        SelectionService,
                        SearchService,
                        PagePreloadService
                    ]
                },] }
    ];
    return GspDocumentModule;
}());
export { GspDocumentModule };
