// @gsp-wf/wf-process-editor/esm5/lib/views/execute-forms/entities/execute-form.entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ExecuteFormModel = /** @class */ (function () {
    function ExecuteFormModel() {
    }
    return ExecuteFormModel;
}());
export { ExecuteFormModel };
if (false) {
    /** @type {?} */
    ExecuteFormModel.prototype.Id;
    /** @type {?} */
    ExecuteFormModel.prototype.Name;
    /** @type {?} */
    ExecuteFormModel.prototype.ComponentId;
    /** @type {?} */
    ExecuteFormModel.prototype.FormType;
    /** @type {?} */
    ExecuteFormModel.prototype.TerminalType;
    /** @type {?} */
    ExecuteFormModel.prototype.ActualParameters;
    /** @type {?} */
    ExecuteFormModel.prototype.formFormatId;
    /** @type {?} */
    ExecuteFormModel.prototype.formatConfigId;
}
var ActualParameter = /** @class */ (function () {
    function ActualParameter() {
    }
    return ActualParameter;
}());
export { ActualParameter };
if (false) {
    /** @type {?} */
    ActualParameter.prototype.Name;
    /** @type {?} */
    ActualParameter.prototype.Value;
    /** @type {?} */
    ActualParameter.prototype.ValueText;
    /** @type {?} */
    ActualParameter.prototype.Tag;
}
/** @enum {string} */
var TerminalType = {
    PC: 'PC',
    MOBILE: 'MOBILE',
    ALL: 'ALL',
};
export { TerminalType };
