// @gsp-wf/wf-process-editor/esm5/lib/views/execute-forms/entities/form-method-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ProcessDeUtil } from '../../../domain/process-de-util';
/** @type {?} */
export var workFlowActions = [
    { text: '正向', value: 'Forward' },
    { text: '反向', value: 'Backward' },
    { text: '终止', value: 'Abort' }
];
/** @type {?} */
export var executionTime = [
    { text: '流转前', value: 'Pre' },
    { text: '流转后', value: 'Post' }
];
var FormMethodConfig = /** @class */ (function () {
    function FormMethodConfig() {
        this.id = ProcessDeUtil.createNM();
        this.workFlowAction = workFlowActions[0].value;
        this.executionTime = executionTime[0].value;
        this.sortOrder = 1;
    }
    return FormMethodConfig;
}());
export { FormMethodConfig };
if (false) {
    /** @type {?} */
    FormMethodConfig.prototype.id;
    /** @type {?} */
    FormMethodConfig.prototype.formFormatConfigId;
    /** @type {?} */
    FormMethodConfig.prototype.workFlowAction;
    /** @type {?} */
    FormMethodConfig.prototype.method;
    /** @type {?} */
    FormMethodConfig.prototype.executionTime;
    /** @type {?} */
    FormMethodConfig.prototype.sortOrder;
    /** @type {?} */
    FormMethodConfig.prototype.methodParameters;
}
