// @gsp-cmp/querysolution/esm5/lib/summary-list/summary-list.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/summary-list/summary-list.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, ElementRef, NgZone, HostBinding, ChangeDetectorRef } from '@angular/core';
import ResizeObserver from "resize-observer-polyfill";
var SummaryListComponent = /** @class */ (function () {
    function SummaryListComponent(ngZone, cd) {
        this.ngZone = ngZone;
        this.cd = cd;
        this.cls = true;
        this.conditions = new Array();
        this.enableTips = false;
        this.ro = null;
        this.widthFlag = 0;
    }
    /**
     * @return {?}
     */
    SummaryListComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.ro) {
            this.ro.unobserve(this.wrapperEl.nativeElement);
            this.ro = null;
        }
    };
    /**
     * @return {?}
     */
    SummaryListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.observerElement();
    };
    /**
     * @private
     * @return {?}
     */
    SummaryListComponent.prototype.observerElement = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            _this.ro = new ResizeObserver((/**
             * @param {?} entries
             * @param {?} observer
             * @return {?}
             */
            function (entries, observer) {
                /** @type {?} */
                var tempWidth = entries[0].contentRect.width;
                // 宽度变化超过80开始计算
                if (Math.abs(tempWidth - _this.widthFlag) > 15) {
                    _this.widthFlag = tempWidth;
                    _this.changeTipsState();
                }
            }));
            _this.ro.observe(_this.wrapperEl.nativeElement);
        }));
    };
    /**
     * @private
     * @return {?}
     */
    SummaryListComponent.prototype.changeTipsState = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var tEnableTips = this.enableTips;
        if (this.wrapperEl.nativeElement.clientWidth + 2 < this.wrapperEl.nativeElement.scrollWidth) {
            tEnableTips = true;
        }
        else {
            tEnableTips = false;
        }
        if (tEnableTips !== this.enableTips) {
            this.enableTips = tEnableTips;
            this.cd.markForCheck();
            this.cd.detectChanges();
        }
    };
    SummaryListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-summary-list',
                    template: "<div class=\"summary-condition-wrapper\" #wrapper [farrisTooltip]=\"enableTips\" [placement]=\"'bottom'\" [content]=\"tpTemplate\" [tipCls]=\"'summary-list--tips'\" [tipWidth]=\"widthFlag+'px'\">\r\n    <ng-container *ngFor=\"let currentItem of conditions;let i = index;let lastIndex = last;\">\r\n        <lib-summary-condition class=\"summary-condition--item\" [condition]=\"currentItem\"></lib-summary-condition>\r\n        <span class=\"summary-condition--spliter\" [ngClass]=\"{'last-spliter': lastIndex}\"></span>\r\n    </ng-container>\r\n</div>\r\n<ng-template #tpTemplate>\r\n    <div class=\"summary-condition--tips\" *ngIf=\"enableTips\">\r\n        <ng-container *ngFor=\"let currentItem of conditions;let i = index;let lastIndex=last;\">\r\n            <lib-summary-condition class=\"summary-condition--item\" [condition]=\"currentItem\"></lib-summary-condition>\r\n            <span class=\"summary-condition--gutter\">;</span>\r\n        </ng-container>\r\n    </div>\r\n</ng-template>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    SummaryListComponent.ctorParameters = function () { return [
        { type: NgZone },
        { type: ChangeDetectorRef }
    ]; };
    SummaryListComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class.summary-list',] }],
        wrapperEl: [{ type: ViewChild, args: ['wrapper', { read: ElementRef },] }],
        conditions: [{ type: Input }],
        enableTips: [{ type: Input }]
    };
    return SummaryListComponent;
}());
export { SummaryListComponent };
if (false) {
    /** @type {?} */
    SummaryListComponent.prototype.cls;
    /** @type {?} */
    SummaryListComponent.prototype.wrapperEl;
    /** @type {?} */
    SummaryListComponent.prototype.conditions;
    /** @type {?} */
    SummaryListComponent.prototype.enableTips;
    /**
     * @type {?}
     * @private
     */
    SummaryListComponent.prototype.ro;
    /** @type {?} */
    SummaryListComponent.prototype.widthFlag;
    /**
     * @type {?}
     * @private
     */
    SummaryListComponent.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    SummaryListComponent.prototype.cd;
}
