// @farris/ide-devkit/esm5/lib/components/panel/sash.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, Output, ViewContainerRef, HostListener, Renderer2 } from '@angular/core';
var SashDirective = /** @class */ (function () {
    function SashDirective(_elementRef, renderer, _viewContainerRef) {
        this._elementRef = _elementRef;
        this.renderer = renderer;
        this._viewContainerRef = _viewContainerRef;
        this.widthChanged = new EventEmitter();
    }
    /**
     * @return {?}
     */
    SashDirective.prototype.onMouseDown = /**
     * @return {?}
     */
    function () {
        console.log('mousedown');
        this.bindDocumentClickListener();
    };
    /**
     * @return {?}
     */
    SashDirective.prototype.onMouseUp = /**
     * @return {?}
     */
    function () {
        console.log('mouseup');
    };
    /**
     * @return {?}
     */
    SashDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    SashDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    SashDirective.prototype.bindDocumentClickListener = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.documentClickListener) {
            this.documentClickListener = this.renderer.listen('document', 'mouseup', function (event) {
                console.log('mouseup');
                console.log(event);
                _this.unbindDocumentMoveListener();
                _this.unbindDocumentClickListener();
            });
            if (!this.documentMoveListener) {
                this.bindDocumentClickListener = this.renderer.listen('document', 'mousemove', function (event) {
                    _this.panelWidth = event.clientX;
                    /** @type {?} */
                    var newWidth = (/** @type {?} */ (event.clientX));
                    _this.widthChanged.emit(newWidth);
                });
            }
        }
    };
    /**
     * @return {?}
     */
    SashDirective.prototype.unbindDocumentClickListener = /**
     * @return {?}
     */
    function () {
        if (this.documentClickListener) {
            this.documentClickListener();
            this.documentClickListener = null;
        }
    };
    /**
     * @return {?}
     */
    SashDirective.prototype.unbindDocumentMoveListener = /**
     * @return {?}
     */
    function () {
        if (this.bindDocumentClickListener) {
            this.bindDocumentClickListener();
            this.bindDocumentClickListener = null;
        }
    };
    SashDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[sash]',
                    exportAs: 'sash'
                },] }
    ];
    /** @nocollapse */
    SashDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: ViewContainerRef }
    ]; };
    SashDirective.propDecorators = {
        panelWidth: [{ type: Input }],
        widthChanged: [{ type: Output }],
        onMouseDown: [{ type: HostListener, args: ['mousedown',] }],
        onMouseUp: [{ type: HostListener, args: ['mouseup',] }]
    };
    return SashDirective;
}());
export { SashDirective };
if (false) {
    /** @type {?} */
    SashDirective.prototype.documentClickListener;
    /** @type {?} */
    SashDirective.prototype.documentMoveListener;
    /** @type {?} */
    SashDirective.prototype.panelWidth;
    /** @type {?} */
    SashDirective.prototype.widthChanged;
    /**
     * @type {?}
     * @private
     */
    SashDirective.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    SashDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    SashDirective.prototype._viewContainerRef;
}
