// @gsp-lcm/metadata-selector/esm2015/metadata/metadata-select/components/metadatapackagelist/metadatapackagelist.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/metadatapackagelist/metadatapackagelist.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { MetadataService } from '../../metadata.service';
import { process } from '@progress/kendo-data-query';
export class MetadatapackagelistComponent {
    /**
     * @param {?} metadataService
     */
    constructor(metadataService) {
        this.metadataService = metadataService;
        // tslint:disable-next-line:member-ordering
        this.state = {
            skip: 0,
            take: 9999,
            // Initial filter descriptor
            filter: {
                logic: 'and',
                filters: []
            }
        };
        // tslint:disable-next-line:member-ordering
        this.selectableSetting = { mode: 'single', checkboxOnly: false };
    }
    /**
     * @param {?} state
     * @return {?}
     */
    dataStateChange(state) {
        this.state = state;
        this.gridData = process(this.data, this.state);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.metadataService.GetMetadataPackages().subscribe((/**
         * @param {?} packages
         * @return {?}
         */
        packages => {
            /** @type {?} */
            const tempdata = [];
            this.packagesInfo = packages;
            if (this.packagesInfo) {
                this.packagesInfo.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => {
                    if (item.type === 'folder' || (item.type === 'file' && item.name.endsWith('mdpkg'))) {
                        tempdata.push({ 'name': item.name });
                    }
                }));
                this.data = tempdata;
                this.gridData = process(this.data, this.state);
            }
        }));
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onSelectionChange(event) {
        /** @type {?} */
        const packageName = event.selectedRows[0].dataItem.name;
        this.metadataService.clearPackage();
        this.metadataService.addPackage(packageName);
    }
}
MetadatapackagelistComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'app-metadatapackagelist',
                template: "<div class=\"templateInfo\">\r\n  <div class=\"templateSelector\">\r\n    <kendo-grid [data]=\"gridData\" \r\n                [selectable]=\"selectableSetting\" \r\n                (selectionChange)=onSelectionChange($event)\r\n                [pageSize]=\"state.take\"\r\n                [skip]=\"state.skip\"\r\n                [sort]=\"state.sort\"\r\n                [filter]=\"state.filter\"\r\n                [sortable]=\"true\"\r\n                [pageable]=\"false\"\r\n                [filterable]=\"true\"\r\n                (dataStateChange)=dataStateChange($event)>\r\n      <kendo-grid-column field=\"name\" title=\"\u5305\u540D\u79F0\" width=\"200\">\r\n      </kendo-grid-column>\r\n    </kendo-grid>\r\n  </div>\r\n</div>",
                styles: [".templateSelector{width:580px;height:450px;border:.5px solid #ddd}.templateDescription{width:200px;height:450px;border:.5px solid #ddd}.templateInfo{display:flex}"]
            }] }
];
/** @nocollapse */
MetadatapackagelistComponent.ctorParameters = () => [
    { type: MetadataService }
];
if (false) {
    /** @type {?} */
    MetadatapackagelistComponent.prototype.data;
    /** @type {?} */
    MetadatapackagelistComponent.prototype.packagesInfo;
    /** @type {?} */
    MetadatapackagelistComponent.prototype.state;
    /** @type {?} */
    MetadatapackagelistComponent.prototype.gridData;
    /** @type {?} */
    MetadatapackagelistComponent.prototype.selectableSetting;
    /**
     * @type {?}
     * @private
     */
    MetadatapackagelistComponent.prototype.metadataService;
}
