// @farris/devkit/esm2015/lib/data-filter/object_filter_condition_data_filter.js
import { Injectable } from '@angular/core';
import { DataFilter } from './data_filter';
export class ObjectFilterConditionDataFilter extends DataFilter {
    constructor() {
        super();
    }
    filter(data, conditions) {
        if (data && data.size) {
            return data.filter((row) => this.validateRowData(row, conditions));
        }
        return data;
    }
    validateRowData(bindingObject, conditions) {
        if (!conditions || Object.keys(conditions).length === 0) {
            return true;
        }
        return this.checkAllFieldInRowData(bindingObject, conditions);
    }
    checkAllFieldInRowData(bindingObject, conditions) {
        let resultObj = null;
        const currentLanguage = this.getCurrentLanguage();
        if (conditions) {
            const fields = Object.keys(conditions);
            resultObj = {};
            fields.forEach((field) => {
                let fieldValue = this.getValue(bindingObject, field);
                const property = bindingObject.properties.find((item) => item.name === field);
                if (property && property.enableMultiLangInput && fieldValue) {
                    fieldValue = fieldValue[currentLanguage];
                }
                const where = conditions[field];
                const { value1, operator1, relation, operator2, value2 } = where;
                if (where) {
                    const result = this.getCompareResult(fieldValue, operator1, value1);
                    resultObj[field] = result;
                    if (relation) {
                        const result2 = this.getCompareResult(fieldValue, operator2, value2);
                        resultObj[field] = this.getRelationResult(result, relation, result2);
                    }
                }
            });
        }
        return this.checkAllFieldResult(resultObj);
    }
    getRelationResult(result1, relation, result2) {
        const _relation = relation.toLowerCase();
        if (_relation === 'and') {
            return result1 && result2;
        }
        else {
            return result1 || result2;
        }
    }
    checkAllFieldResult(resultObj) {
        if (!resultObj) {
            return true;
        }
        return Object.values(resultObj).reduce((c, n) => {
            return c && n;
        }, true);
    }
}
ObjectFilterConditionDataFilter.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ObjectFilterConditionDataFilter.ctorParameters = () => [];
