// @farris/mobile-vue-adapter/esm2015/lib/component-extenders/map-extender.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component-extenders/map-extender.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 列表组件扩展
 */
class MapExtender {
    /**
     * @param {?} mapInstance
     * @param {?} viewModelContext
     */
    constructor(mapInstance, viewModelContext) {
        this.mapInstance = mapInstance;
        this.viewModelContext = viewModelContext;
    }
    /**
     * 扩展组件
     * @return {?}
     */
    extend() {
        /** @type {?} */
        const self = this;
        // 当前行
        this.mapInstance.$on('markMap', (/**
         * @param {?} selectedData
         * @return {?}
         */
        (selectedData) => {
            // 获取经纬度
            /** @type {?} */
            const lnglat = selectedData.lnglat;
            /** @type {?} */
            const lng = lnglat.lng;
            /** @type {?} */
            const lat = lnglat.lat;
            // 获取经纬度路径
            /** @type {?} */
            const lngPath = this.mapInstance.lngPath;
            /** @type {?} */
            const latPath = this.mapInstance.latPath;
            // this.mapInstance.center = [lng, lat];
            // 赋值经纬度
            this.viewModelContext.bindingData.setValue(this.getBindingPathArray().concat(lngPath.split('/')), lng, true, true);
            this.viewModelContext.bindingData.setValue(this.getBindingPathArray().concat(latPath.split('/')), lat, true, true);
        }));
    }
    /**
     * 清理状态
     * @return {?}
     */
    dispose() {
        this.mapInstance.$off('markMap');
    }
    /**
     * @private
     * @return {?}
     */
    getBindingPathArray() {
        /** @type {?} */
        const path = this.viewModelContext.bindingData.bindingPath;
        if (path) {
            return path.split('/').filter((/**
             * @param {?} n
             * @return {?}
             */
            (n) => n !== ''));
        }
        return [];
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    MapExtender.prototype.mapInstance;
    /**
     * @type {?}
     * @private
     */
    MapExtender.prototype.viewModelContext;
}
export { MapExtender };
