// @farris/mobile-vue-adapter/esm2015/lib/component-extenders/reflesh-extender.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component-extenders/reflesh-extender.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 列表组件扩展
 */
class RefleshExtender {
    /**
     * @param {?} refleshInstance
     * @param {?} viewModelContext
     */
    constructor(refleshInstance, viewModelContext) {
        this.refleshInstance = refleshInstance;
        this.viewModelContext = viewModelContext;
    }
    /**
     * 扩展组件
     * @return {?}
     */
    extend() {
        /** @type {?} */
        const self = this;
        // 监听下拉刷新组件
        this.refleshInstance.$on('input', (/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            if (data) {
                self.viewModelContext.repository.setPaginationInfo({
                    pageIndex: 1
                });
                self.viewModelContext.uiState.setPropertyValue('listaviewState', Object.assign({}, self.viewModelContext.uiState['listaviewState'], { loadMoreLoading: true }));
                self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
            }
            self.viewModelContext.appContext.eventBus.subscribe(`${this.viewModelContext.id}-reflesh-extend`, (/**
             * @param {?} value
             * @return {?}
             */
            (value) => {
                if (value.type === 'loadDataFinished') {
                    self.viewModelContext.uiState.setPropertyValue('listaviewState', Object.assign({}, self.viewModelContext.uiState['listaviewState'], { refreshLoading: false, loadMoreLoading: false }));
                }
            }));
        }));
    }
    /**
     * 清理状态
     * @return {?}
     */
    dispose() {
        this.refleshInstance.$off('input');
        this.viewModelContext.appContext.eventBus.unsubscribe(`${this.viewModelContext.id}-reflesh-extend`);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    RefleshExtender.prototype.refleshInstance;
    /**
     * @type {?}
     * @private
     */
    RefleshExtender.prototype.viewModelContext;
}
export { RefleshExtender };
