// @farris/mobile-vue-adapter/esm2015/lib/component-extenders/search-extender.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component-extenders/search-extender.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 列表组件扩展
 */
class SearchExtender {
    /**
     * @param {?} searchInstance
     * @param {?} viewModelContext
     */
    constructor(searchInstance, viewModelContext) {
        this.searchInstance = searchInstance;
        this.viewModelContext = viewModelContext;
    }
    /**
     * 扩展组件
     * @return {?}
     */
    extend() {
        /** @type {?} */
        const self = this;
        // 监听搜索组件
        this.searchInstance.$on('cancel', (/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            //  清空走查询
            self.viewModelContext.repository.setPaginationInfo({
                pageIndex: 1
            });
            self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
        }));
        this.searchInstance.$on('search', (/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            // 如果是确定类型，即走查询
            self.viewModelContext.repository.setPaginationInfo({
                pageIndex: 1
            });
            self.viewModelContext.appContext.eventBus.triggerEvent({ type: 'LoadData' });
        }));
    }
    /**
     * 清理状态
     * @return {?}
     */
    dispose() {
        this.searchInstance.$off('cancel');
        this.searchInstance.$off('search');
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    SearchExtender.prototype.searchInstance;
    /**
     * @type {?}
     * @private
     */
    SearchExtender.prototype.viewModelContext;
}
export { SearchExtender };
