// @gsp-cmp/common-component/esm5/convertor/data-type/structure-refer-type-convertor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var StructureReferTypeConvertor = /** @class */ (function () {
    function StructureReferTypeConvertor() {
    }
    /**
     * @param {?} obj
     * @param {?} datatype
     * @param {?} ctxt
     * @return {?}
     */
    StructureReferTypeConvertor.prototype.convertToObject = /**
     * @param {?} obj
     * @param {?} datatype
     * @param {?} ctxt
     * @return {?}
     */
    function (obj, datatype, ctxt) {
        /** @type {?} */
        var referType = (/** @type {?} */ (datatype));
        obj["kind"] = referType.getKind();
        obj['referId'] = referType.referId;
        obj['code'] = referType.code;
        obj['name'] = referType.name;
    };
    /**
     * @param {?} dataType
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    StructureReferTypeConvertor.prototype.initFromObject = /**
     * @param {?} dataType
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    function (dataType, obj, ctxt) {
        /** @type {?} */
        var refType = (/** @type {?} */ (dataType));
        refType.kind = obj['kind'];
        refType.referId = obj['referId'];
        refType.code = obj['code'];
        refType.name = obj['name'];
    };
    return StructureReferTypeConvertor;
}());
export { StructureReferTypeConvertor };
