// @gsp-wf/wf-process-editor/esm5/lib/views/activity-component/activity-component.viewmodel.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ActivityComponentUIState } from './activity-component.uistate';
import { ActivityComponentRepository } from './activity-component.repository';
import { MessagerService } from '@farris/ui-messager';
import { WfLocalePipe } from '../../i18n/pipe/wf-process-editor-locale.pipe';
var ActivityComponentViewModel = /** @class */ (function () {
    function ActivityComponentViewModel(uistate, repository, messager, localePipe) {
        this.uistate = uistate;
        this.repository = repository;
        this.messager = messager;
        this.localePipe = localePipe;
    }
    /**
     * @return {?}
     */
    ActivityComponentViewModel.prototype.deleteRow = /**
     * @return {?}
     */
    function () {
        if (this.uistate.curComponent) {
            if (this.uistate.curComponent.executionTime === 'LOADDATA') {
                this.messager.info(this.localePipe.transform('component.activityList.name'));
            }
            else {
                this.repository.removeComponent(this.uistate.curComponent.id);
            }
        }
        else {
            return;
        }
    };
    /**
     * @param {?} cpt
     * @return {?}
     */
    ActivityComponentViewModel.prototype.componentSelect = /**
     * @param {?} cpt
     * @return {?}
     */
    function (cpt) {
        if (cpt) {
            this.repository.addComponent(cpt);
        }
    };
    /**
     * @param {?} flowFormId
     * @param {?} bizActId
     * @return {?}
     */
    ActivityComponentViewModel.prototype.loadComponents = /**
     * @param {?} flowFormId
     * @param {?} bizActId
     * @return {?}
     */
    function (flowFormId, bizActId) {
        this.repository.loadComponents(flowFormId, bizActId);
    };
    ActivityComponentViewModel.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ActivityComponentViewModel.ctorParameters = function () { return [
        { type: ActivityComponentUIState },
        { type: ActivityComponentRepository },
        { type: MessagerService },
        { type: WfLocalePipe }
    ]; };
    return ActivityComponentViewModel;
}());
export { ActivityComponentViewModel };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ActivityComponentViewModel.prototype.uistate;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentViewModel.prototype.repository;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentViewModel.prototype.messager;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentViewModel.prototype.localePipe;
}
