// @gsp-wf/rtdevkit/esm5/lib/runtime/components/user-helper/user-helper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Optional, Output, EventEmitter } from '@angular/core';
import { UserService } from '../../services/user.service';
import { DataTableComponent } from '@farris/ui-datatable';
import { NotifyService } from '@farris/ui-notify';
import { DialogComponent } from '@farris/ui-dialog';
import { I18nService } from '../../services/i18n.service';
var UserHelperComponent = /** @class */ (function () {
    function UserHelperComponent(i18nService, userSvc, notify) {
        this.i18nService = i18nService;
        this.userSvc = userSvc;
        this.notify = notify;
        this.pageList = [10, 20, 30, 50, 100];
        this.paginationOptions = {
            id: 'WF-Users-DataTable-Pagination',
            currentPage: 1,
            itemsPerPage: 10,
            pageList: this.pageList,
            totalItems: 0
        };
        this.columns = [
            { title: this.i18nService ? this.i18nService.getValue('static.columnName.code') : '编号', field: 'Code', width: 80 },
            { title: this.i18nService ? this.i18nService.getValue('static.columnName.name') : '姓名', field: 'Name', width: 90 },
            { title: this.i18nService ? this.i18nService.getValue('static.columnName.department') : '部门', field: 'OrgName', width: 200 }
        ];
        this.searchValue = null;
        this.userListData = [];
        // 是否允许多选,默认多选
        this.singleSelect = false;
        this.selectedUsers = [];
        this.completed = new EventEmitter();
    }
    /**
     * @return {?}
     */
    UserHelperComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    UserHelperComponent.prototype.handleOk = /**
     * @return {?}
     */
    function () {
        throw new Error('没有实现，需要子组件实现');
    };
    /**
     * @return {?}
     */
    UserHelperComponent.prototype.handleCancel = /**
     * @return {?}
     */
    function () {
        this.completed.emit(false);
        this.showModal(false);
    };
    /**
     * @param {?} pageSize
     * @return {?}
     */
    UserHelperComponent.prototype.onPageSizeChange = /**
     * @param {?} pageSize
     * @return {?}
     */
    function (pageSize) {
        this.paginationOptions.itemsPerPage = pageSize;
        this.getUserList();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    UserHelperComponent.prototype.onPageChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.paginationOptions.currentPage = event.pageIndex;
        this.getUserList();
    };
    /**
     * @protected
     * @return {?}
     */
    UserHelperComponent.prototype.getUserList = /**
     * @protected
     * @return {?}
     */
    function () {
        var _this = this;
        this.userSvc.getUserList(this.searchValue, this.paginationOptions.currentPage, this.paginationOptions.itemsPerPage)
            .subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            _this.paginationOptions.totalItems = re.totalCount;
            _this.userListData = re.data.map((/**
             * @param {?} u
             * @return {?}
             */
            function (u) { return ({
                ID: u.id,
                Code: u.code,
                Name: u.name,
                OrgId: u.sysOrgId,
                OrgName: u.sysOrgName,
                Checked: false
            }); }));
        }));
    };
    /**
     * @param {?} event
     * @return {?}
     */
    UserHelperComponent.prototype.search = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.paginationOptions.currentPage = 1;
        this.searchValue = encodeURIComponent(event.value);
        if (event.field === '*') {
            this.getUserList();
        }
        else if (event.field === 'Name') {
            this.getUserList();
        }
        else {
            this.notify.warning('只支持用户名称过滤！');
        }
    };
    /**
     * @protected
     * @param {?} show
     * @return {?}
     */
    UserHelperComponent.prototype.showModal = /**
     * @protected
     * @param {?} show
     * @return {?}
     */
    function (show) {
        if (show) {
            this.userDialog.show();
            this.getUserList();
        }
        else {
            // 临时处理方案
            if (this.userDialog.beforeClose) {
                this.userDialog.close();
            }
        }
    };
    UserHelperComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-user-helper',
                    template: "<farris-dialog #allUserDialog [buttons]=\"buttonRef\" [width]=\"850\" [height]=\"520\" [title]=\"'\u7CFB\u7EDF\u7528\u6237'\">\r\n  <div class=\"d-flex flex-column my-data-grid h-100\" style=\"padding: 0 14px;\">\r\n    <div class=\"flex-auto\" style=\"position:relative;\">\r\n      <farris-datatable #allUserTable [columns]=\"columns\" [data]=\"userListData | paginate: paginationOptions\"\r\n        [showFilterBar]='true' [singleSelect]=\"singleSelect\" [idField]=\"'ID'\" [remote]=\"'server'\" (search)=\"search($event)\"\r\n        [pagination]='false'></farris-datatable>\r\n    </div>\r\n    <div style=\"position:relative;\">\r\n      <pagination-controls [id]=\"paginationOptions.id\" [previousLabel]=\"' '\" [nextLabel]=\"' '\"\r\n        (pageChange)=\"onPageChange($event)\" (pageSizeChange)=\"onPageSizeChange($event)\">\r\n      </pagination-controls>\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n<ng-template #buttonRef>\r\n  <button class=\"btn btn-sm btn-primary px-3\" (click)=\"handleOk()\">\u786E\u5B9A</button>\r\n  <button class=\"btn btn-sm btn-default px-3\" (click)=\"handleCancel()\">\u53D6\u6D88</button>\r\n</ng-template>",
                    styles: [".my-data-grid .filter-row .search-fields{padding-left:0!important}.my-data-grid .filter-row .search-text{padding-right:0!important}"]
                }] }
    ];
    /** @nocollapse */
    UserHelperComponent.ctorParameters = function () { return [
        { type: I18nService, decorators: [{ type: Optional }] },
        { type: UserService },
        { type: NotifyService }
    ]; };
    UserHelperComponent.propDecorators = {
        userTable: [{ type: ViewChild, args: ['allUserTable',] }],
        userDialog: [{ type: ViewChild, args: ['allUserDialog',] }],
        completed: [{ type: Output }]
    };
    return UserHelperComponent;
}());
export { UserHelperComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    UserHelperComponent.prototype.pageList;
    /** @type {?} */
    UserHelperComponent.prototype.paginationOptions;
    /** @type {?} */
    UserHelperComponent.prototype.columns;
    /** @type {?} */
    UserHelperComponent.prototype.searchValue;
    /** @type {?} */
    UserHelperComponent.prototype.userListData;
    /** @type {?} */
    UserHelperComponent.prototype.singleSelect;
    /** @type {?} */
    UserHelperComponent.prototype.selectedUsers;
    /** @type {?} */
    UserHelperComponent.prototype.userTable;
    /** @type {?} */
    UserHelperComponent.prototype.userDialog;
    /** @type {?} */
    UserHelperComponent.prototype.completed;
    /**
     * @type {?}
     * @protected
     */
    UserHelperComponent.prototype.i18nService;
    /**
     * @type {?}
     * @protected
     */
    UserHelperComponent.prototype.userSvc;
    /**
     * @type {?}
     * @protected
     */
    UserHelperComponent.prototype.notify;
}
