// @gsp-svc/cloudprint/esm5/lib/cloudprint.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { LOCALE_ID, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { CloudprintDesignerComponent } from './cloudprint-designer/cloudprint-designer.component';
import { CloudprintViewerComponent } from './cloudprint-viewer/cloudprint-viewer.component';
import { CloudprintdesignerService } from './cloudprint-designer/cloudprint-designer.service';
import { CloudprintviewerService } from './cloudprint-viewer/cloudprint-viewer.service';
import { CloudprintService } from './cloudprint.service';
import { NotifyService, NotifyContainerComponent, NotifyModule } from '@farris/ui-notify';
import { CloudprintFunctionRegister } from './functions/cloudprint.function.register';
import { AppService } from '@gsp-sys/rtf-common';
import { PrintFormatService } from './services/printFormatService';
import { FormatlistComponent } from './cloudprint-formatlist/formatlist.component';
import { WindowModule, DialogModule } from '@progress/kendo-angular-dialog';
import { FormsModule } from '@angular/forms';
import { GridModule } from '@progress/kendo-angular-grid';
import { LanguageService } from '@ecp-caf/caf-common';
import { FormatlistjobComponent } from './cloudprint-formatlistjob/formatlistjob.component';
import { LoadingModule, LoadingService } from '@farris/ui-loading';
import { FarrisButtonModule } from '@farris/ui-button';
import { DatagridModule } from '@farris/ui-datagrid';
import { SplitterModule } from '@farris/ui-splitter';
import { FarrisSectionModule } from '@farris/ui-section';
import { AngularDraggableModule } from '@farris/ui-draggable';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { ModalModule } from '@farris/ui-modal';
import { NumberSpinnerModule } from '@farris/ui-number-spinner';
import { ComboListModule } from '@farris/ui-combo-list';
import { FarrisFormsModule } from '@farris/ui-forms';
import { LocalLangPipe } from './local.pipe';
// import { BrowserModule } from '@angular/platform-browser';
import { FResponseToolbarModule } from '@farris/ui-response-toolbar';
var ɵ0 = localStorage.getItem('languageCode') || 'zh-CHS';
var CloudPrintModule = /** @class */ (function () {
    function CloudPrintModule() {
    }
    CloudPrintModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        // BrowserModule,
                        HttpClientModule,
                        NotifyModule,
                        WindowModule,
                        FormsModule,
                        GridModule,
                        DialogModule,
                        LoadingModule,
                        LoadingModule.forRoot(),
                        FarrisDialogModule,
                        FarrisButtonModule,
                        SplitterModule,
                        DatagridModule,
                        AngularDraggableModule,
                        FarrisSectionModule,
                        ModalModule.forRoot(),
                        NumberSpinnerModule,
                        ComboListModule,
                        FarrisFormsModule,
                        FResponseToolbarModule
                    ],
                    declarations: [
                        CloudprintDesignerComponent,
                        CloudprintViewerComponent,
                        FormatlistComponent,
                        FormatlistjobComponent,
                        LocalLangPipe
                    ],
                    entryComponents: [
                        CloudprintDesignerComponent,
                        CloudprintViewerComponent,
                        NotifyContainerComponent,
                        FormatlistComponent,
                        FormatlistjobComponent
                    ],
                    providers: [
                        // component里依赖的服务都需要在这里申明下
                        CloudprintdesignerService,
                        CloudprintviewerService,
                        CloudprintService,
                        NotifyService,
                        CloudprintFunctionRegister,
                        AppService,
                        PrintFormatService,
                        LanguageService,
                        LoadingService,
                        { provide: LOCALE_ID, useValue: ɵ0 },
                    ],
                },] }
    ];
    return CloudPrintModule;
}());
export { CloudPrintModule };
export { ɵ0 };
