// @gsp-wf/wf-process-model/esm2015/lib/model/AifCreationBillActivity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, BpmnModelJsonHelper, StartEvent } from "@edp-pmf/bpmn-model";
export class AifCreationBillActivity extends StartEvent {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = AifCreationBillActivity.Clr_Type_ID;
        this.aifProcessCode = null;
        this.voNodePath = null;
        this.isRelied = true;
        this.aifProcessCode = null;
        this.voNodePath = null;
        this.isRelied = true;
    }
    /**
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.aifProcessCode = null;
        this.voNodePath = null;
        this.isRelied = true;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, 'aifProcessCode', this.aifProcessCode);
        BpmnModelHelper.SetAttributeValue(obj, 'voNodePath', this.voNodePath);
        BpmnModelHelper.SetAttributeValue(obj, 'isRelied', this.isRelied);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.aifProcessCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "aifProcessCode");
        this.voNodePath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voNodePath");
        this.isRelied = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isRelied");
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new AifCreationBillActivity(this.Model);
    }
    /**
     * 模型校验
     * @param {?} errors 模型校验错误信息
     * @return {?}
     */
    Validate(errors) {
        // TODO: 需要添加自定义校验信息
        super.Validate(errors);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const creationAct = new AifCreationBillActivity(this.Model);
        creationAct.Name = this.Name;
        creationAct.aifProcessCode = null;
        if (this.aifProcessCode) {
            creationAct.aifProcessCode = this.aifProcessCode;
        }
        creationAct.voNodePath = null;
        if (this.voNodePath) {
            creationAct.voNodePath = this.voNodePath;
        }
        creationAct.isRelied = this.isRelied;
        return creationAct;
    }
}
AifCreationBillActivity.Clr_Type_ID = "wf.AifCreationBillActivity";
if (false) {
    /** @type {?} */
    AifCreationBillActivity.Clr_Type_ID;
    /** @type {?} */
    AifCreationBillActivity.prototype.ClrTypeID;
    /** @type {?} */
    AifCreationBillActivity.prototype.aifProcessCode;
    /** @type {?} */
    AifCreationBillActivity.prototype.voNodePath;
    /** @type {?} */
    AifCreationBillActivity.prototype.isRelied;
}
BpmnModelHelper.RegisterElementType(new AifCreationBillActivity(new BpmnModel()));
