// @gsp-wf/wf-process-model/esm2015/lib/model/ApprovalPivotConfig.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionMethod, ExtensionElement } from "@edp-pmf/bpmn-model";
import { ApprovalPivot } from "./ApprovalPivot";
export class ApprovalPivotConfig extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = ApprovalPivotConfig.Clr_Type_ID;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.approvalPivots = new Array();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "Title", this.title);
        ExtensionMethod.SetAttributeValue(obj, "Enable", this.enable);
        ExtensionMethod.SetAttributeValue(obj, "Expand", this.expand);
        /** @type {?} */
        const approvalPivots = new Array();
        if (this.approvalPivots !== null && this.approvalPivots.length > 0) {
            this.approvalPivots.forEach((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                approvalPivots.push(item.ConvertToJsonObject());
            }));
        }
        ExtensionMethod.SetAttributeValue(obj, "ApprovalPivots", approvalPivots);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.title = ExtensionMethod.GetAttributeValue4String(obj, "Title");
        this.enable = ExtensionMethod.GetAttributeValue4Boolean(obj, "Enable");
        this.expand = ExtensionMethod.GetAttributeValue4Boolean(obj, "Expand");
        /** @type {?} */
        const approvalPivots = (/** @type {?} */ (ExtensionMethod.GetAttributeValue(obj, "ApprovalPivots")));
        if (approvalPivots !== null && approvalPivots.length > 0) {
            approvalPivots.forEach((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                /** @type {?} */
                const approvalPivot = new ApprovalPivot(this.Model);
                approvalPivot.LoadFromJsonObject(item);
                this.approvalPivots.push(approvalPivot);
            }));
        }
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ApprovalPivotConfig(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const approvalPivotConfig = new ApprovalPivotConfig(this.Model);
        approvalPivotConfig.title = this.title;
        approvalPivotConfig.enable = this.enable;
        approvalPivotConfig.expand = this.expand;
        approvalPivotConfig.approvalPivots = new Array();
        if (this.approvalPivots && this.approvalPivots.length > 0) {
            for (const p of this.approvalPivots) {
                /** @type {?} */
                const approvalPivot = p.clone();
                approvalPivotConfig.approvalPivots.push(approvalPivot);
            }
        }
        return approvalPivotConfig;
    }
}
ApprovalPivotConfig.Clr_Type_ID = "wf.ApprovalPivotConfig";
if (false) {
    /** @type {?} */
    ApprovalPivotConfig.Clr_Type_ID;
    /** @type {?} */
    ApprovalPivotConfig.prototype.ClrTypeID;
    /** @type {?} */
    ApprovalPivotConfig.prototype.title;
    /** @type {?} */
    ApprovalPivotConfig.prototype.enable;
    /** @type {?} */
    ApprovalPivotConfig.prototype.expand;
    /** @type {?} */
    ApprovalPivotConfig.prototype.approvalPivots;
}
BpmnModelHelper.RegisterElementType(new ApprovalPivotConfig(new BpmnModel()));
