// @gsp-wf/wf-process-model/esm2015/lib/model/ApprovalPivot.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionMethod, ExtensionElement } from "@edp-pmf/bpmn-model";
import { Condition } from "./extension/condition-expression/Condition";
export class ApprovalPivot extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = ApprovalPivot.Clr_Type_ID;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "Type", this.type);
        ExtensionMethod.SetAttributeValue(obj, "Title", this.title);
        ExtensionMethod.SetAttributeValue(obj, "Description", this.description);
        if (this.condition) {
            BpmnModelHelper.SetAttributeValue(obj, "Condition", this.condition.ConvertToJsonObject());
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.type = (/** @type {?} */ (ExtensionMethod.GetAttributeValue4String(obj, "Type")));
        this.title = ExtensionMethod.GetAttributeValue4String(obj, "Title");
        this.description = ExtensionMethod.GetAttributeValue4String(obj, "Description");
        /** @type {?} */
        const conditionElement = ExtensionMethod.GetAttributeValue(obj, 'Condition');
        if (conditionElement) {
            /** @type {?} */
            const condition = new Condition(this.Model);
            condition.LoadFromJsonObject(conditionElement);
            this.condition = condition;
        }
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ApprovalPivot(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const approvalPivot = new ApprovalPivot(this.Model);
        approvalPivot.type = this.type;
        approvalPivot.title = this.title;
        approvalPivot.description = this.description;
        if (this.condition) {
            approvalPivot.condition = this.condition.clone();
        }
        return approvalPivot;
    }
}
ApprovalPivot.Clr_Type_ID = "wf.ApprovalPivot";
if (false) {
    /** @type {?} */
    ApprovalPivot.Clr_Type_ID;
    /** @type {?} */
    ApprovalPivot.prototype.ClrTypeID;
    /** @type {?} */
    ApprovalPivot.prototype.type;
    /** @type {?} */
    ApprovalPivot.prototype.condition;
    /** @type {?} */
    ApprovalPivot.prototype.title;
    /** @type {?} */
    ApprovalPivot.prototype.description;
}
BpmnModelHelper.RegisterElementType(new ApprovalPivot(new BpmnModel()));
