// @gsp-wf/wf-process-model/esm2015/lib/model/AutoHandleExpireOperation.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { ExpireOperation } from "./ExpireOperation";
export class AutoHandleExpireOperation extends ExpireOperation {
    constructor() {
        super(...arguments);
        this.ClrTypeID = AutoHandleExpireOperation.Clr_Type_ID;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "Message", this.message);
        ExtensionMethod.SetAttributeValue(obj, "AutoHandleType", this.autoHandleType);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.message = ExtensionMethod.GetAttributeValue4String(obj, "Message");
        this.autoHandleType = ExtensionMethod.GetAttributeValue4String(obj, "AutoHandleType");
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new AutoHandleExpireOperation(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const autoHandleExpireOperation = new AutoHandleExpireOperation(this.Model);
        autoHandleExpireOperation.id = BpmnModelHelper.GenerateElementId();
        autoHandleExpireOperation.durationType = this.durationType;
        autoHandleExpireOperation.timeDuration = this.timeDuration;
        autoHandleExpireOperation.timeUnit = this.timeUnit;
        autoHandleExpireOperation.message = this.message;
        autoHandleExpireOperation.autoHandleType = this.autoHandleType;
        return autoHandleExpireOperation;
    }
}
AutoHandleExpireOperation.Clr_Type_ID = "wf.Mex.AutoHandleExpireOperation";
if (false) {
    /** @type {?} */
    AutoHandleExpireOperation.Clr_Type_ID;
    /** @type {?} */
    AutoHandleExpireOperation.prototype.ClrTypeID;
    /**
     * 办理类型，通过驳回否决等等
     * @type {?}
     */
    AutoHandleExpireOperation.prototype.autoHandleType;
    /**
     * 办理意见
     * @type {?}
     */
    AutoHandleExpireOperation.prototype.message;
}
BpmnModelHelper.RegisterElementType(new AutoHandleExpireOperation(new BpmnModel()));
