// @gsp-wf/wf-process-model/esm2015/lib/model/CheckCondition.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { ConditionValueVersion } from "./extension/condition-expression/ConditionValueVersion";
import { MexExpressionImpl } from "./mex/MexExpressionImpl";
export class CheckCondition extends MexExpressionImpl {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = CheckCondition.Clr_Type_ID;
        this.valueVersion = ConditionValueVersion.Unsure;
        this.checkTime = 'UNDEFINITION';
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.value = '';
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "CheckTime", this.checkTime);
        ExtensionMethod.SetAttributeValue(obj, "ValueVersion", this.valueVersion);
        if (this.errorTips && this.errorTips.size > 0) {
            /** @type {?} */
            const jsonArray = new Array();
            this.errorTips.forEach((/**
             * @param {?} value
             * @param {?} key
             * @return {?}
             */
            (value, key) => {
                /** @type {?} */
                const childObj = new Object;
                ExtensionMethod.SetAttributeValue(childObj, "Key", key);
                ExtensionMethod.SetAttributeValue(childObj, "Value", value);
                jsonArray.push(childObj);
            }));
            ExtensionMethod.SetAttributeValue(obj, "ErrorTips", jsonArray);
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.checkTime = ExtensionMethod.GetAttributeValue4String(obj, "CheckTime");
        this.valueVersion = (/** @type {?} */ (ExtensionMethod.GetAttributeValue4String(obj, "ValueVersion")));
        /** @type {?} */
        const jsonArray = ExtensionMethod.GetAttributeValue(obj, "ErrorTips");
        if (jsonArray && jsonArray.length > 0) {
            if (!this.errorTips) {
                this.errorTips = new Map();
            }
            for (const childObj of jsonArray) {
                /** @type {?} */
                const key = ExtensionMethod.GetAttributeValue4String(childObj, "Key");
                /** @type {?} */
                const value = ExtensionMethod.GetAttributeValue4String(childObj, "Value");
                this.errorTips.set(key, value);
            }
        }
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const condition = new CheckCondition(this.Model);
        condition.value = this.value;
        condition.valueVersion = this.valueVersion;
        condition.checkTime = this.checkTime;
        if (this.errorTips && this.errorTips.size > 0) {
            condition.errorTips = new Map();
            this.errorTips.forEach((/**
             * @param {?} value
             * @param {?} key
             * @return {?}
             */
            (value, key) => {
                condition.errorTips.set(key, value);
            }));
        }
        return condition;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new CheckCondition(this.Model);
    }
}
CheckCondition.Clr_Type_ID = "wf.CheckCondition";
if (false) {
    /** @type {?} */
    CheckCondition.Clr_Type_ID;
    /** @type {?} */
    CheckCondition.prototype.ClrTypeID;
    /** @type {?} */
    CheckCondition.prototype.valueVersion;
    /** @type {?} */
    CheckCondition.prototype.checkTime;
    /** @type {?} */
    CheckCondition.prototype.errorTips;
}
BpmnModelHelper.RegisterElementType(new CheckCondition(new BpmnModel()));
