// @gsp-wf/wf-process-model/esm2015/lib/model/ExpireOperation.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ExtensionMethod, MexExtensionElement } from "@edp-pmf/bpmn-model";
/**
 * @abstract
 */
export class ExpireOperation extends MexExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "Id", this.id);
        ExtensionMethod.SetAttributeValue(obj, "DurationType", this.durationType);
        ExtensionMethod.SetAttributeValue(obj, "TimeDuration", this.timeDuration);
        ExtensionMethod.SetAttributeValue(obj, "TimeUnit", this.timeUnit);
        ExtensionMethod.SetAttributeValue(obj, "OperationTypeName", this.operationTypeName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.id = ExtensionMethod.GetAttributeValue4String(obj, "Id");
        this.durationType = (/** @type {?} */ (ExtensionMethod.GetAttributeValue4String(obj, "DurationType")));
        this.timeDuration = ExtensionMethod.GetAttributeValue4Int64(obj, "TimeDuration");
        this.timeUnit = ExtensionMethod.GetAttributeValue4String(obj, "TimeUnit");
        this.operationTypeName = ExtensionMethod.GetAttributeValue4String(obj, "OperationTypeName");
    }
}
if (false) {
    /** @type {?} */
    ExpireOperation.prototype.id;
    /**
     * 期限控制方式
     * @type {?}
     */
    ExpireOperation.prototype.durationType;
    /**
     * 任务期限
     * @type {?}
     */
    ExpireOperation.prototype.timeDuration;
    /**
     * 时间单位
     * @type {?}
     */
    ExpireOperation.prototype.timeUnit;
    /**
     * 操作类型名称
     * @type {?}
     */
    ExpireOperation.prototype.operationTypeName;
}
