// @gsp-wf/wf-process-model/esm2015/lib/model/ExtendProperty.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
export class ExtendProperty extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = ExtendProperty.Clr_Type_ID;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "Key", this.key);
        ExtensionMethod.SetAttributeValue(obj, "Value", this.value);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.key = ExtensionMethod.GetAttributeValue4String(obj, "Key");
        this.value = ExtensionMethod.GetAttributeValue4String(obj, "Value");
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ExtendProperty(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const extendProperty = new ExtendProperty(this.Model);
        extendProperty.key = this.key;
        extendProperty.value = this.value;
        return extendProperty;
    }
}
ExtendProperty.Clr_Type_ID = "wf.ExtendProperty";
if (false) {
    /** @type {?} */
    ExtendProperty.Clr_Type_ID;
    /** @type {?} */
    ExtendProperty.prototype.ClrTypeID;
    /**
     * 扩展属性key值
     * @type {?}
     */
    ExtendProperty.prototype.key;
    /**
     * 扩展属性value值
     * @type {?}
     */
    ExtendProperty.prototype.value;
}
BpmnModelHelper.RegisterElementType(new ExtendProperty(new BpmnModel()));
