// @gsp-wf/wf-process-model/esm2015/lib/model/ImageSetting.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
export class ImageSettings extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = ImageSettings.Clr_Type_ID;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "ImageOperation", this.imageOperation);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.imageOperation = ExtensionMethod.GetAttributeValue4String(obj, "ImageOperation");
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ImageSettings(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const imageSettings = new ImageSettings(this.Model);
        imageSettings.imageOperation = this.imageOperation;
        return imageSettings;
    }
}
ImageSettings.Clr_Type_ID = "wf.ImageSettings";
if (false) {
    /** @type {?} */
    ImageSettings.Clr_Type_ID;
    /** @type {?} */
    ImageSettings.prototype.ClrTypeID;
    /**
     * 影像操作权限
     * @type {?}
     */
    ImageSettings.prototype.imageOperation;
}
BpmnModelHelper.RegisterElementType(new ImageSettings(new BpmnModel()));
