// @gsp-wf/wf-process-model/esm2015/lib/model/MessageNotice.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { ActivityExecutorParticipantExpression } from "./participants/ActivityExecutorParticipantExpression";
import { ComponentParticipantExpression } from "./participants/ComponentParticipantExpression";
import { MatrixParticipantExpression } from "./participants/MatrixParticipantExpression";
import { OrganizationParticipantExpression } from "./participants/OrganizationParticipantExpression";
import { PositionParticipantExpression } from "./participants/PositionParticipantExpression";
import { ReportingLineParticipantExpression } from "./participants/ReportingLineParticipantExpression";
import { UserParticipantExpression } from "./participants/UserParticipantExpression";
import { SmsMessageTemplate } from "./SmsMessageTemplate";
export class MessageNotice extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = MessageNotice.Clr_Type_ID;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.receiverExpressions = new Array();
        this.smsMessageTemplate = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "Content", this.content);
        ExtensionMethod.SetAttributeValue(obj, "SendTime", this.sendTime);
        ExtensionMethod.SetAttributeValue(obj, "SendType", this.sendType);
        ExtensionMethod.SetAttributeValue(obj, "Subject", this.subject);
        ExtensionMethod.SetAttributeValue(obj, "Receiver", this.receiver);
        ExtensionMethod.SetAttributeValue(obj, "ReceiverText", this.receiverText);
        if (this.receiverExpressions != null && this.receiverExpressions.length > 0) {
            /** @type {?} */
            const receiverList = new Array();
            for (const receiver of this.receiverExpressions) {
                receiverList.push(receiver.ConvertToJsonObject());
            }
            ExtensionMethod.SetAttributeValue(obj, "ReceiverExpressions", receiverList);
        }
        if (this.smsMessageTemplate) {
            ExtensionMethod.SetAttributeValue(obj, "SmsMessageTemplate", this.smsMessageTemplate.ConvertToJsonObject());
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.content = ExtensionMethod.GetAttributeValue4String(obj, "Content");
        this.sendTime = ExtensionMethod.GetAttributeValue4String(obj, "SendTime");
        this.sendType = ExtensionMethod.GetAttributeValue4String(obj, "SendType");
        this.subject = ExtensionMethod.GetAttributeValue4String(obj, "Subject");
        this.receiver = ExtensionMethod.GetAttributeValue4String(obj, "Receiver");
        this.receiverText = ExtensionMethod.GetAttributeValue4String(obj, "ReceiverText");
        /** @type {?} */
        const receiverList = ExtensionMethod.GetAttributeValue(obj, "ReceiverExpressions");
        if (receiverList && receiverList.length > 0) {
            this.receiverExpressions = new Array();
            for (const receiver of receiverList) {
                /** @type {?} */
                const clrTypeID = ExtensionMethod.GetAttributeValue4String(receiver, "ClrTypeID");
                switch (clrTypeID) {
                    case UserParticipantExpression.Clr_Type_ID:
                        /** @type {?} */
                        const userParticipantExpression = new UserParticipantExpression(this.Model);
                        userParticipantExpression.LoadFromJsonObject(receiver);
                        this.receiverExpressions.push(userParticipantExpression);
                        break;
                    case PositionParticipantExpression.Clr_Type_ID:
                        /** @type {?} */
                        const positionParticipantExpression = new PositionParticipantExpression(this.Model);
                        positionParticipantExpression.LoadFromJsonObject(receiver);
                        this.receiverExpressions.push(positionParticipantExpression);
                        break;
                    case ReportingLineParticipantExpression.Clr_Type_ID:
                        /** @type {?} */
                        const reportingLineParticipantExpression = new ReportingLineParticipantExpression(this.Model);
                        reportingLineParticipantExpression.LoadFromJsonObject(receiver);
                        this.receiverExpressions.push(reportingLineParticipantExpression);
                        break;
                    case OrganizationParticipantExpression.Clr_Type_ID:
                        /** @type {?} */
                        const organizationParticipantExpression = new OrganizationParticipantExpression(this.Model);
                        organizationParticipantExpression.LoadFromJsonObject(receiver);
                        this.receiverExpressions.push(organizationParticipantExpression);
                        break;
                    case MatrixParticipantExpression.Clr_Type_ID:
                        /** @type {?} */
                        const matrixParticipantExpression = new MatrixParticipantExpression(this.Model);
                        matrixParticipantExpression.LoadFromJsonObject(receiver);
                        this.receiverExpressions.push(matrixParticipantExpression);
                        break;
                    case ComponentParticipantExpression.Clr_Type_ID:
                        /** @type {?} */
                        const componentParticipantExpression = new ComponentParticipantExpression(this.Model);
                        componentParticipantExpression.LoadFromJsonObject(receiver);
                        this.receiverExpressions.push(componentParticipantExpression);
                        break;
                    case ActivityExecutorParticipantExpression.Clr_Type_ID:
                        /** @type {?} */
                        const activityExecutorParticipantExpression = new ActivityExecutorParticipantExpression(this.Model);
                        activityExecutorParticipantExpression.LoadFromJsonObject(receiver);
                        this.receiverExpressions.push(activityExecutorParticipantExpression);
                        break;
                    default:
                        break;
                }
            }
        }
        /** @type {?} */
        const smsTemplateElement = ExtensionMethod.GetAttributeValue(obj, 'SmsMessageTemplate');
        if (smsTemplateElement) {
            /** @type {?} */
            const smsMessageTemplate = new SmsMessageTemplate(this.Model);
            smsMessageTemplate.LoadFromJsonObject(smsTemplateElement);
            this.smsMessageTemplate = smsMessageTemplate;
        }
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new MessageNotice(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const messageNotice = new MessageNotice(this.Model);
        messageNotice.content = this.content;
        messageNotice.sendTime = this.sendTime;
        messageNotice.sendType = this.sendType;
        messageNotice.subject = this.subject;
        messageNotice.receiver = this.receiver;
        messageNotice.receiverText = this.receiverText;
        messageNotice.receiverExpressions = new Array();
        if (this.receiverExpressions && this.receiverExpressions.length > 0) {
            for (const r of this.receiverExpressions) {
                if (r instanceof OrganizationParticipantExpression) {
                    /** @type {?} */
                    const o = (/** @type {?} */ (r));
                    /** @type {?} */
                    const organizationParticipantExpression = o.clone();
                    messageNotice.receiverExpressions.push(organizationParticipantExpression);
                }
                else if (r instanceof UserParticipantExpression) {
                    /** @type {?} */
                    const u = (/** @type {?} */ (r));
                    /** @type {?} */
                    const userParticipantExpression = u.clone();
                    messageNotice.receiverExpressions.push(userParticipantExpression);
                }
                else if (r instanceof PositionParticipantExpression) {
                    /** @type {?} */
                    const p = (/** @type {?} */ (r));
                    /** @type {?} */
                    const positionParticipantExpression = p.clone();
                    messageNotice.receiverExpressions.push(positionParticipantExpression);
                }
                else if (r instanceof ReportingLineParticipantExpression) {
                    /** @type {?} */
                    const re = (/** @type {?} */ (r));
                    /** @type {?} */
                    const reportingLineParticipantExpression = re.clone();
                    messageNotice.receiverExpressions.push(reportingLineParticipantExpression);
                }
                else if (r instanceof MatrixParticipantExpression) {
                    /** @type {?} */
                    const m = (/** @type {?} */ (r));
                    /** @type {?} */
                    const matrixParticipantExpression = m.clone();
                    messageNotice.receiverExpressions.push(matrixParticipantExpression);
                }
                else if (r instanceof ComponentParticipantExpression) {
                    /** @type {?} */
                    const c = (/** @type {?} */ (r));
                    /** @type {?} */
                    const componentParticipantExpression = c.clone();
                    messageNotice.receiverExpressions.push(componentParticipantExpression);
                }
                else if (r instanceof ActivityExecutorParticipantExpression) {
                    /** @type {?} */
                    const a = (/** @type {?} */ (r));
                    /** @type {?} */
                    const activityExecutorParticipantExpression = a.clone();
                    messageNotice.receiverExpressions.push(activityExecutorParticipantExpression);
                }
            }
        }
        if (this.smsMessageTemplate) {
            messageNotice.smsMessageTemplate = this.smsMessageTemplate.clone();
        }
        return messageNotice;
    }
}
MessageNotice.Clr_Type_ID = "wf.MessageNotice";
if (false) {
    /** @type {?} */
    MessageNotice.Clr_Type_ID;
    /** @type {?} */
    MessageNotice.prototype.ClrTypeID;
    /**
     * 消息内容
     * @type {?}
     */
    MessageNotice.prototype.content;
    /**
     * 消息发送时机
     * @type {?}
     */
    MessageNotice.prototype.sendTime;
    /**
     * 消息发送方式
     * @type {?}
     */
    MessageNotice.prototype.sendType;
    /**
     * 消息主题
     * @type {?}
     */
    MessageNotice.prototype.subject;
    /**
     * 消息接收人
     * @type {?}
     */
    MessageNotice.prototype.receiver;
    /**
     * 消息接收人显示
     * @type {?}
     */
    MessageNotice.prototype.receiverText;
    /** @type {?} */
    MessageNotice.prototype.receiverExpressions;
    /** @type {?} */
    MessageNotice.prototype.smsMessageTemplate;
}
BpmnModelHelper.RegisterElementType(new MessageNotice(new BpmnModel()));
