// @gsp-wf/wf-process-model/esm2015/lib/model/SignalEventDefinitionImpl.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionMethod, SignalEventDefinition } from "@edp-pmf/bpmn-model";
export class SignalEventDefinitionImpl extends SignalEventDefinition {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = SignalEventDefinitionImpl.Clr_Type_ID;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "Name", this.name);
        ExtensionMethod.SetAttributeValue(obj, "TriggerAction", this.triggerAction);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.name = ExtensionMethod.GetAttributeValue4String(obj, "Name");
        this.triggerAction = ExtensionMethod.GetAttributeValue4String(obj, "TriggerAction");
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new SignalEventDefinitionImpl(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const signalEventDefinitionImpl = new SignalEventDefinitionImpl(this.Model);
        signalEventDefinitionImpl.name = this.name;
        signalEventDefinitionImpl.triggerAction = this.triggerAction;
        signalEventDefinitionImpl.SignalExpression = this.SignalExpression;
        signalEventDefinitionImpl.SignalRef = this.SignalRef;
        signalEventDefinitionImpl.Async = this.Async;
        return signalEventDefinitionImpl;
    }
}
SignalEventDefinitionImpl.Clr_Type_ID = "wf.SignalEventDefinitionImpl";
if (false) {
    /** @type {?} */
    SignalEventDefinitionImpl.Clr_Type_ID;
    /** @type {?} */
    SignalEventDefinitionImpl.prototype.ClrTypeID;
    /** @type {?} */
    SignalEventDefinitionImpl.prototype.name;
    /** @type {?} */
    SignalEventDefinitionImpl.prototype.triggerAction;
}
BpmnModelHelper.RegisterElementType(new SignalEventDefinitionImpl(new BpmnModel()));
