// @gsp-wf/wf-process-model/esm2015/lib/model/SmsMessageTemplate.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { SmsMessageVariable } from "./SmsMessageVariable";
export class SmsMessageTemplate extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = SmsMessageTemplate.Clr_Type_ID;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.smsMessageVariables = new Array();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "Code", this.code);
        ExtensionMethod.SetAttributeValue(obj, "Content", this.content);
        if (this.smsMessageVariables && this.smsMessageVariables.length > 0) {
            /** @type {?} */
            const variables = new Array();
            for (const variable of this.smsMessageVariables) {
                variables.push(variable.ConvertToJsonObject());
            }
            BpmnModelHelper.SetAttributeValue(obj, "SmsMessageVariables", variables);
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.code = ExtensionMethod.GetAttributeValue4String(obj, "Code");
        this.content = ExtensionMethod.GetAttributeValue4String(obj, "Content");
        this.smsMessageVariables = new Array();
        /** @type {?} */
        const variables = ExtensionMethod.GetAttributeValue(obj, "SmsMessageVariables");
        if (variables && variables.length > 0) {
            for (const variable of variables) {
                /** @type {?} */
                const smsMessageVariable = new SmsMessageVariable(this.Model);
                smsMessageVariable.LoadFromJsonObject(variable);
                this.smsMessageVariables.push(smsMessageVariable);
            }
        }
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new SmsMessageTemplate(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const smsMessageTemplate = new SmsMessageTemplate(this.Model);
        smsMessageTemplate.code = this.code;
        smsMessageTemplate.content = this.content;
        smsMessageTemplate.smsMessageVariables = new Array();
        if (this.smsMessageVariables && this.smsMessageVariables.length > 0) {
            for (const variable of this.smsMessageVariables) {
                /** @type {?} */
                const smsMessageVariable = variable.clone();
                smsMessageTemplate.smsMessageVariables.push(smsMessageVariable);
            }
        }
        return smsMessageTemplate;
    }
}
SmsMessageTemplate.Clr_Type_ID = "wf.SmsMessageTemplate";
if (false) {
    /** @type {?} */
    SmsMessageTemplate.Clr_Type_ID;
    /** @type {?} */
    SmsMessageTemplate.prototype.ClrTypeID;
    /** @type {?} */
    SmsMessageTemplate.prototype.code;
    /** @type {?} */
    SmsMessageTemplate.prototype.content;
    /** @type {?} */
    SmsMessageTemplate.prototype.smsMessageVariables;
}
BpmnModelHelper.RegisterElementType(new SmsMessageTemplate(new BpmnModel()));
