// @gsp-wf/wf-process-model/esm2015/lib/model/SmsMessageVariable.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
export class SmsMessageVariable extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = SmsMessageVariable.Clr_Type_ID;
        this.tag = 'expression';
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "Code", this.code);
        ExtensionMethod.SetAttributeValue(obj, "Value", this.value);
        ExtensionMethod.SetAttributeValue(obj, "Tag", this.tag);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.code = ExtensionMethod.GetAttributeValue4String(obj, "Code");
        this.value = ExtensionMethod.GetAttributeValue4String(obj, "Value");
        this.tag = ExtensionMethod.GetAttributeValue4String(obj, "Tag");
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new SmsMessageVariable(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const smsMessageVariable = new SmsMessageVariable(this.Model);
        smsMessageVariable.code = this.code;
        smsMessageVariable.value = this.value;
        smsMessageVariable.tag = this.tag;
        return smsMessageVariable;
    }
}
SmsMessageVariable.Clr_Type_ID = "wf.SmsMessageVariable";
if (false) {
    /** @type {?} */
    SmsMessageVariable.Clr_Type_ID;
    /** @type {?} */
    SmsMessageVariable.prototype.ClrTypeID;
    /** @type {?} */
    SmsMessageVariable.prototype.code;
    /** @type {?} */
    SmsMessageVariable.prototype.value;
    /** @type {?} */
    SmsMessageVariable.prototype.tag;
}
BpmnModelHelper.RegisterElementType(new SmsMessageVariable(new BpmnModel()));
