// @gsp-wf/wf-process-model/esm2015/lib/model/TimeLimitControl.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionMethod, ExtensionElement } from "@edp-pmf/bpmn-model";
import { AutoHandleExpireOperation } from "./AutoHandleExpireOperation";
import { SendMessageExpireOperation } from "./SendMessageExpireOperation";
export class TimeLimitControl extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = TimeLimitControl.Clr_Type_ID;
        this.expireOperations = new Array();
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "Tag", this.tag);
        ExtensionMethod.SetAttributeValue(obj, "DueTime", this.dueTime);
        ExtensionMethod.SetAttributeValue(obj, "TimeUnit", this.timeUnit);
        ExtensionMethod.SetAttributeValue(obj, "Enable", this.enable);
        /** @type {?} */
        const expireOperationElements = new Array();
        if (this.expireOperations !== null && this.expireOperations.length > 0) {
            this.expireOperations.forEach((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                expireOperationElements.push(item.ConvertToJsonObject());
            }));
        }
        ExtensionMethod.SetAttributeValue(obj, "ExpireOperations", expireOperationElements);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.tag = ExtensionMethod.GetAttributeValue4String(obj, "Tag");
        this.dueTime = ExtensionMethod.GetAttributeValue4String(obj, "DueTime");
        this.timeUnit = ExtensionMethod.GetAttributeValue4String(obj, "TimeUnit");
        this.enable = ExtensionMethod.GetAttributeValue4Boolean(obj, "Enable");
        /** @type {?} */
        const expireOperationElements = (/** @type {?} */ (ExtensionMethod.GetAttributeValue(obj, "ExpireOperations")));
        if (expireOperationElements !== null && expireOperationElements.length > 0) {
            expireOperationElements.forEach((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                if (item.OperationTypeName === 'SendMessageExpireOperation') {
                    /** @type {?} */
                    const expireOperation = new SendMessageExpireOperation(this.Model);
                    expireOperation.LoadFromJsonObject(item);
                    this.expireOperations.push(expireOperation);
                }
                else if (item.OperationTypeName === 'AutoHandleExpireOperation') {
                    /** @type {?} */
                    const expireOperation = new AutoHandleExpireOperation(this.Model);
                    expireOperation.LoadFromJsonObject(item);
                    this.expireOperations.push(expireOperation);
                }
            }));
        }
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new TimeLimitControl(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const timeLimitControl = new TimeLimitControl(this.Model);
        timeLimitControl.tag = this.tag;
        timeLimitControl.enable = this.enable;
        timeLimitControl.dueTime = this.dueTime;
        timeLimitControl.timeUnit = this.timeUnit;
        timeLimitControl.expireOperations = new Array();
        if (this.expireOperations && this.expireOperations.length > 0) {
            for (const e of this.expireOperations) {
                if (e instanceof SendMessageExpireOperation) {
                    /** @type {?} */
                    const sendMessage = (/** @type {?} */ (e));
                    /** @type {?} */
                    const expireOperation = sendMessage.clone();
                    timeLimitControl.expireOperations.push(expireOperation);
                }
            }
        }
        return timeLimitControl;
    }
}
TimeLimitControl.Clr_Type_ID = "wf.TimeLimitControl";
if (false) {
    /** @type {?} */
    TimeLimitControl.Clr_Type_ID;
    /** @type {?} */
    TimeLimitControl.prototype.ClrTypeID;
    /**
     * 自定义标签，用于区分期限控制方式
     * expression 表达式
     * constant 固定值
     * scope 时间范围
     * @type {?}
     */
    TimeLimitControl.prototype.tag;
    /**
     * 是否启用
     * @type {?}
     */
    TimeLimitControl.prototype.enable;
    /**
     * 限定时长
     * @type {?}
     */
    TimeLimitControl.prototype.dueTime;
    /**
     * 时间单位 day,hour,minute
     * @type {?}
     */
    TimeLimitControl.prototype.timeUnit;
    /** @type {?} */
    TimeLimitControl.prototype.expireOperations;
}
BpmnModelHelper.RegisterElementType(new TimeLimitControl(new BpmnModel()));
