// @gsp-wf/wf-process-model/esm2015/lib/model/ValuedDataObject.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, DataObject } from "@edp-pmf/bpmn-model";
import { VariableSetting } from "./VariableSetting";
export class ValuedDataObject extends DataObject {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = ValuedDataObject.Clr_Type_ID;
    }
    /**
     * @protected
     * @return {?}
     */
    GetExtensionElements4ConvertTo() {
        /** @type {?} */
        let list = super.GetExtensionElements4ConvertTo();
        if (this.variableSetting != null) {
            this.variableSetting.MexTag = ValuedDataObject.MEXTAG;
            list.push(this.variableSetting);
        }
        return list;
    }
    /**
     * @protected
     * @param {?} extensionElements
     * @return {?}
     */
    SetExtensionElements4LoadFrom(extensionElements) {
        super.SetExtensionElements4LoadFrom(extensionElements);
        for (let extensionElement of extensionElements) {
            if (extensionElement.MexTag !== ValuedDataObject.MEXTAG)
                continue;
            switch (extensionElement.ClrTypeID) {
                case VariableSetting.Clr_Type_ID:
                    this.variableSetting = ((/** @type {?} */ (extensionElement)));
                    break;
                default:
                    break;
            }
        }
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ValuedDataObject(this.Model);
    }
}
ValuedDataObject.Clr_Type_ID = "wf.ValuedDataObject";
ValuedDataObject.MEXTAG = 'wf.ValuedDataObject';
if (false) {
    /** @type {?} */
    ValuedDataObject.Clr_Type_ID;
    /**
     * @type {?}
     * @protected
     */
    ValuedDataObject.MEXTAG;
    /** @type {?} */
    ValuedDataObject.prototype.ClrTypeID;
    /** @type {?} */
    ValuedDataObject.prototype.variableSetting;
}
BpmnModelHelper.RegisterElementType(new ValuedDataObject(new BpmnModel()));
