// @gsp-wf/wf-process-model/esm2015/lib/model/VariableSetting.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionMethod, ExtensionElement } from "@edp-pmf/bpmn-model";
export class VariableSetting extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = VariableSetting.Clr_Type_ID;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        // ExtensionMethod.SetAttributeValue(obj, "VariableId", this.variableId);
        ExtensionMethod.SetAttributeValue(obj, "VariableCode", this.variableCode);
        ExtensionMethod.SetAttributeValue(obj, "VariableName", this.variableName);
        ExtensionMethod.SetAttributeValue(obj, "Tag", this.tag);
        ExtensionMethod.SetAttributeValue(obj, "ExecutionTime", this.executionTime);
        ExtensionMethod.SetAttributeValue(obj, "Value", this.value);
        ExtensionMethod.SetAttributeValue(obj, "ValueText", this.valueText);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        // this.variableId = ExtensionMethod.GetAttributeValue4String(obj, "VariableId");
        this.variableCode = ExtensionMethod.GetAttributeValue4String(obj, "VariableCode");
        this.variableName = ExtensionMethod.GetAttributeValue4String(obj, "VariableName");
        this.tag = ExtensionMethod.GetAttributeValue4String(obj, "Tag");
        this.executionTime = ExtensionMethod.GetAttributeValue4String(obj, "ExecutionTime");
        this.value = ExtensionMethod.GetAttributeValue4String(obj, "Value");
        this.valueText = ExtensionMethod.GetAttributeValue4String(obj, "ValueText");
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new VariableSetting(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const variableSetting = new VariableSetting(this.Model);
        variableSetting.variableCode = this.variableCode;
        variableSetting.variableName = this.variableName;
        variableSetting.tag = this.tag;
        variableSetting.executionTime = this.executionTime;
        variableSetting.value = this.value;
        variableSetting.valueText = this.valueText;
        return variableSetting;
    }
}
VariableSetting.Clr_Type_ID = "wf.VariableSetting";
if (false) {
    /** @type {?} */
    VariableSetting.Clr_Type_ID;
    /** @type {?} */
    VariableSetting.prototype.ClrTypeID;
    /**
     * 变量编号
     * @type {?}
     */
    VariableSetting.prototype.variableCode;
    /**
     * 变量名称
     * @type {?}
     */
    VariableSetting.prototype.variableName;
    /**
     * 自定义标签，赋值方式
     * expression 语义化表达式
     * constant 常量
     * component 构件
     * @type {?}
     */
    VariableSetting.prototype.tag;
    /**
     * 变量执行时机
     * ACTIVITY_STARTED 活动开始
     * ACTIVITY_COMPLETED 活动完成
     * @type {?}
     */
    VariableSetting.prototype.executionTime;
    /**
     * 变量值
     * @type {?}
     */
    VariableSetting.prototype.value;
    /** @type {?} */
    VariableSetting.prototype.valueText;
}
BpmnModelHelper.RegisterElementType(new VariableSetting(new BpmnModel()));
