// @farris/ide-devkit/esm2015/lib/components/ng2tree-common/components/tree-node-drop-slot/tree-node-drop-slot.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { TreeNode } from '../../models/tree-node';
export class TreeNodeDropSlotComponent {
    /**
     * @param {?} $event
     * @return {?}
     */
    onDragOver($event) {
        $event.preventDefault();
        this.node.treeModel.setDropLocation({ component: this, parentNode: this.node, index: this.dropIndex });
    }
    /**
     * @return {?}
     */
    onDragLeave() {
        if (this.node.treeModel.isDraggingOver(this)) {
            this.node.treeModel.setDropLocation(null);
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onDrop($event) {
        $event.preventDefault();
        // this.node.mouseAction('drop', $event, { node: this.node, index: this.dropIndex });
        console.log('onDropEvent: tree-node-drop-slot');
        this.node.dropMouseAction($event, { parentNode: this.node, index: this.dropIndex });
    }
}
TreeNodeDropSlotComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng2tree-node-drop-slot',
                template: "<div\r\n  class=\"node-drop-slot\"\r\n  [class.is-dragging-over]=\"node.treeModel.isDraggingOver(this)\"\r\n  (drop)=\"onDrop($event)\"\r\n  (dragover)=\"onDragOver($event)\"\r\n  (dragleave)=\"onDragLeave()\"\r\n>\r\n</div>",
                styles: [".node-drop-slot{display:block;height:2px;width:100%}.node-drop-slot.is-dragging-over{background:#dfe;height:20px;border:2px dotted #888}"]
            }] }
];
TreeNodeDropSlotComponent.propDecorators = {
    node: [{ type: Input }],
    dropIndex: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TreeNodeDropSlotComponent.prototype.node;
    /** @type {?} */
    TreeNodeDropSlotComponent.prototype.dropIndex;
}
