// @farris/ui-search-box/esm2015/lib/search-panel/search-panel.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ComponentFactoryResolver, Directive, EventEmitter, HostListener, Injector, Input, NgZone, Output, Renderer2 } from '@angular/core';
import { reqAnimFrame } from '@farris/ui-common';
import { InputGroupComponent } from '@farris/ui-input-group';
import { debounceTime } from 'rxjs/operators';
import { SearchBoxComponent } from '../search-box.component';
import { trim } from 'lodash-es';
import { SearchPanelComponent } from './search-panel.component';
export class SearchPanelDirective {
    /**
     * @param {?} injector
     * @param {?} ngzone
     * @param {?} render
     * @param {?} inputRef
     * @param {?} searchBoxRef
     * @param {?} cfr
     */
    constructor(injector, ngzone, render, inputRef, searchBoxRef, cfr) {
        this.injector = injector;
        this.ngzone = ngzone;
        this.render = render;
        this.inputRef = inputRef;
        this.searchBoxRef = searchBoxRef;
        this.cfr = cfr;
        this.maxFields = 5;
        /**
         * 启用 任意字段
         */
        this.useAnyField = true;
        this.escHandler = new EventEmitter();
        this.maxPanelItems = 10;
        this.minPanelWidth = 270;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if ((!this.fields || !this.fields.length) && !this.useAnyField) {
            return;
        }
        this.inputRef.inputClick.subscribe((/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            e.stopPropagation();
        }));
        this.inputRef.valueChange.pipe(debounceTime(50)).subscribe((/**
         * @param {?} val
         * @return {?}
         */
        (val) => {
            if (trim(val)) {
                // 值变化显示查询面板
                if (!this.panelElement) {
                    this.createDataPanel();
                    this.searchPanelRef.instance.setActiveItem(0);
                }
                this.searchPanelRef.instance.setValue(val);
            }
            else {
                if (this.panelElement) {
                    this.hide();
                }
            }
        }));
        this.searchBoxRef.clear.subscribe((/**
         * @return {?}
         */
        () => {
            this.hide();
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.searchBoxRef.overLayService.destory(this.searchBoxRef.el.nativeElement);
    }
    /**
     * @private
     * @return {?}
     */
    removePanelElement() {
        document.body.removeChild(this.panelElement);
        this.panelElement = null;
        if (this.searchPanelRef) {
            this.searchPanelRef.destroy();
            this.searchPanelRef = null;
        }
    }
    /**
     * @param {?=} e
     * @return {?}
     */
    hide(e) {
        reqAnimFrame((/**
         * @return {?}
         */
        () => {
            this.inputRef.value = '';
            if (this.panelElement) {
                if (e && (e.type === 'mousewheel' || e.type === 'wheel')) {
                    this.removePanelElement();
                }
                else {
                    this.panelElement.classList.remove('f-area-show');
                    setTimeout((/**
                     * @return {?}
                     */
                    () => {
                        this.removePanelElement();
                    }), 120);
                }
                this.searchBoxRef.overLayService.destory(this.searchBoxRef.el.nativeElement);
            }
        }));
    }
    /**
     * @private
     * @param {?=} updateTopPosition
     * @return {?}
     */
    setPanelPosition(updateTopPosition = true) {
        if (this.panelElement) {
            const { maxWidth, left, top, height } = this.getPanelSize();
            this.panelElement.style.minWidth = `${this.minPanelWidth}px`;
            this.panelElement.style.maxWidth = `${maxWidth}px`;
            this.panelElement.style.maxHeight = `${height}px`;
            if (updateTopPosition) {
                this.panelElement.style.top = `${top}px`;
            }
            this.panelElement.style.left = `${left}px`;
        }
    }
    /**
     * @private
     * @return {?}
     */
    createDataPanel() {
        this.panelElement = document.createElement('div');
        this.panelElement.classList.add('overlay-pane', 'f-search-box-panel', 'f-area-hide');
        /** @type {?} */
        const zindex = '' + this.searchBoxRef.commonUtils.getFloatingLayerIndex();
        document.body.appendChild(this.panelElement);
        this.setPanelPosition();
        this.panelElement.style.zIndex = zindex;
        this.panelElement.style.overflow = 'hidden';
        // 创建数据展示组件
        /** @type {?} */
        const cmpFact = this.cfr.resolveComponentFactory(SearchPanelComponent);
        this.searchPanelRef = cmpFact.create(this.injector);
        if (this.useAnyField && !this.fields.find((/**
         * @param {?} n
         * @return {?}
         */
        n => n.code === '*'))) {
            /** @type {?} */
            const anyfieldsName = this.searchBoxRef.localeSer.getValue('searchbox.anyFields');
            this.fields.splice(0, 0, { code: '*', name: anyfieldsName });
        }
        this.searchPanelRef.instance.fields = this.fields;
        this.panelElement.appendChild(this.searchPanelRef.location.nativeElement);
        this.searchPanelRef.changeDetectorRef.detectChanges();
        this.searchPanelRef.instance.searchKeyWord = this.inputRef.value;
        this.searchPanelRef.instance.itemClick.subscribe((/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            const { data } = e;
            this.selectItem(data);
        }));
        this.searchPanelRef.instance.closepanel.subscribe((/**
         * @return {?}
         */
        () => {
            this.hide();
        }));
        // 注册鼠标滚轮，点击事件，用于隐藏Panel
        this.searchBoxRef.overLayService.registerMouseEvent(this.searchBoxRef.el.nativeElement, (/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            if (!this.panelElement || e.target['closest']('.f-search-box-panel')) {
                return;
            }
            if (this.inputRef && this.inputRef.textbox.nativeElement === e.target) {
                return false;
            }
            // 输入框内有文本时，如果不允许关闭弹出面板，可放开
            // if (this.inputRef.value) {
            //     return;
            // }
            this.hide(e);
        }));
        if (this.searchPanelRef.instance.fields.length + 1 < this.maxPanelItems) {
            /** @type {?} */
            let newHeight = this.searchPanelRef.instance.itemsContainer.nativeElement.querySelector('ul').offsetHeight + 40;
            this.searchPanelRef.instance.itemsContainer.nativeElement.style.height = `${newHeight}px`;
            if (this.panelElement) {
                if (this.panelElement.style.transformOrigin.indexOf('bottom') > -1) {
                    this.panelElement.style.top = `${this.getInputSizeInfo().top - newHeight - 5}px`;
                }
                this.panelElement.style.height = 'auto';
            }
        }
        else {
            this.searchPanelRef.instance.itemsContainer.nativeElement.style.height = '100%';
            if (this.panelElement) {
                this.render.removeStyle(this.panelElement, 'height');
            }
        }
        this.panelElement.classList.add('f-area-show');
    }
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    selectItem(data) {
        if (!data) {
            return;
        }
        this.searchBoxRef.select(Object.assign({}, data, { value: this.inputRef.value }));
        this.hide();
        this.inputRef.focus();
    }
    /**
     * @private
     * @return {?}
     */
    calculationPanelHeight() {
        return this.maxPanelItems * 30 + 15;
    }
    /**
     * @private
     * @return {?}
     */
    getInputSizeInfo() {
        /** @type {?} */
        const el = this.searchBoxRef.enableFloat ? this.searchBoxRef.shadowBox : this.searchBoxRef.container;
        return el.nativeElement.getBoundingClientRect();
    }
    /**
     * @private
     * @return {?}
     */
    getPanelSize() {
        let { width, height, top, left } = this.getInputSizeInfo();
        /** @type {?} */
        const bottom = window.innerHeight - height - top;
        /** @type {?} */
        let panelHeight = this.calculationPanelHeight();
        /** @type {?} */
        const h = top > bottom ? top : bottom;
        if (bottom > panelHeight) {
            top = top + height;
            // 面板由上向下展开
            this.panelElement.style.transformOrigin = '100% top';
        }
        else {
            if (top > bottom) {
                if (h < panelHeight) {
                    panelHeight = h - 10;
                    top = 10;
                }
                else {
                    top = top - parseInt('' + panelHeight, 10) - 5;
                }
                // 面板由下向上展开
                this.panelElement.style.transformOrigin = '100% bottom';
            }
            else {
                if (h < panelHeight) {
                    panelHeight = h - 10;
                }
                top = top + height;
            }
        }
        /** @type {?} */
        let _width = width < this.minPanelWidth ? this.minPanelWidth : width;
        if (window.innerWidth - left < _width) {
            left = left + width - _width;
        }
        return { width: _width, top, height: panelHeight, left, maxWidth: Math.floor(window.innerWidth - left - 10) };
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    registerKeyboardEvent($event) {
        /** @type {?} */
        const rows = this.fields;
        if ($event.code === 'ArrowUp' || $event.code === 'ArrowDown') {
            $event.preventDefault();
            $event.stopPropagation();
            if (this.searchPanelRef) {
                /** @type {?} */
                const idx = this.searchPanelRef.instance.activeIndex;
                /** @type {?} */
                const setActiveItem = (/**
                 * @param {?} index
                 * @return {?}
                 */
                (index) => {
                    this.searchPanelRef.instance.setActiveItem(index);
                });
                if ($event.code === 'ArrowUp') { // up
                    if (idx > -1) {
                        /** @type {?} */
                        let prevIdx = idx - 1;
                        if (prevIdx < 0) {
                            prevIdx = rows.length - 1;
                        }
                        setActiveItem(prevIdx);
                    }
                    else {
                        setActiveItem(rows.length - 1);
                    }
                }
                if ($event.code === 'ArrowDown') { // down
                    // down
                    /** @type {?} */
                    let nextIdx = idx + 1;
                    if (nextIdx >= rows.length) {
                        nextIdx = 0;
                    }
                    setActiveItem(nextIdx);
                }
            }
        }
        if ($event.code === 'Backspace') {
            if (!this.inputRef.value) {
                if (this.searchBoxRef.displayTextList && this.searchBoxRef.displayTextList.length) {
                    $event.preventDefault();
                    $event.stopPropagation();
                    /** @type {?} */
                    const removeItem = this.searchBoxRef.displayTextList[this.searchBoxRef.displayTextList.length - 1];
                    if (removeItem) {
                        this.searchBoxRef.remove(removeItem.code);
                        if (this.searchPanelRef) {
                            this.searchPanelRef.changeDetectorRef.detectChanges();
                        }
                    }
                }
            }
        }
        if ($event.key === 'Enter') {
            if (rows && rows.length && this.panelElement) {
                /** @type {?} */
                const idx = this.searchPanelRef.instance.activeIndex;
                /** @type {?} */
                const data = rows[idx];
                this.selectItem(data);
            }
            else {
                this.searchBoxRef.onSubmit($event);
            }
        }
        if ($event.key === 'Escape') {
            if (this.panelElement) {
                this.hide();
            }
            else {
                this.escHandler.emit();
            }
        }
    }
}
SearchPanelDirective.decorators = [
    { type: Directive, args: [{ selector: '[search-box-panel]' },] }
];
/** @nocollapse */
SearchPanelDirective.ctorParameters = () => [
    { type: Injector },
    { type: NgZone },
    { type: Renderer2 },
    { type: InputGroupComponent },
    { type: SearchBoxComponent },
    { type: ComponentFactoryResolver }
];
SearchPanelDirective.propDecorators = {
    fields: [{ type: Input, args: ['search-box-panel',] }],
    maxFields: [{ type: Input }],
    useAnyField: [{ type: Input }],
    escHandler: [{ type: Output }],
    registerKeyboardEvent: [{ type: HostListener, args: ['keydown', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    SearchPanelDirective.prototype.fields;
    /** @type {?} */
    SearchPanelDirective.prototype.maxFields;
    /**
     * 启用 任意字段
     * @type {?}
     */
    SearchPanelDirective.prototype.useAnyField;
    /** @type {?} */
    SearchPanelDirective.prototype.escHandler;
    /**
     * @type {?}
     * @private
     */
    SearchPanelDirective.prototype.panelElement;
    /**
     * @type {?}
     * @private
     */
    SearchPanelDirective.prototype.searchPanelRef;
    /**
     * @type {?}
     * @private
     */
    SearchPanelDirective.prototype.maxPanelItems;
    /**
     * @type {?}
     * @private
     */
    SearchPanelDirective.prototype.minPanelWidth;
    /**
     * @type {?}
     * @private
     */
    SearchPanelDirective.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    SearchPanelDirective.prototype.ngzone;
    /**
     * @type {?}
     * @private
     */
    SearchPanelDirective.prototype.render;
    /**
     * @type {?}
     * @private
     */
    SearchPanelDirective.prototype.inputRef;
    /**
     * @type {?}
     * @private
     */
    SearchPanelDirective.prototype.searchBoxRef;
    /**
     * @type {?}
     * @private
     */
    SearchPanelDirective.prototype.cfr;
}
