// @farris/ui-search-box/esm2015/lib/search-panel/search-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, HostListener, Injector, Input, Output, ViewChild, ViewEncapsulation } from '@angular/core';
export class SearchPanelComponent {
    /**
     * @param {?} injector
     * @param {?} cdr
     */
    constructor(injector, cdr) {
        this.injector = injector;
        this.cdr = cdr;
        this.fields = [];
        this.maxItems = 5;
        this.searchKeyWord = '';
        this.itemClick = new EventEmitter();
        this.closepanel = new EventEmitter();
        this.mouseDown = new EventEmitter();
        this.activeIndex = -1;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClick($event) {
        $event.stopPropagation();
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @param {?} items
     * @return {?}
     */
    loadData(items) {
        this.fields = items;
        if (!this.cdr['destroyed']) {
            this.cdr.detectChanges();
        }
    }
    /**
     * @param {?} $event
     * @param {?} item
     * @return {?}
     */
    onFieldItemClick($event, item) {
        $event.preventDefault();
        this.itemClick.emit({ data: item, evnet: $event });
    }
    /**
     * @param {?} index
     * @return {?}
     */
    setActiveItem(index) {
        this.activeIndex = index;
        if (!this.cdr['destroyed']) {
            this.cdr.detectChanges();
        }
    }
    /**
     * @param {?} val
     * @return {?}
     */
    setValue(val) {
        this.searchKeyWord = val;
        this.cdr.detectChanges();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClosePanel($event) {
        $event.stopPropagation();
        this.closepanel.emit();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onMouseDown($event) {
        $event.preventDefault();
        $event.stopPropagation();
        this.mouseDown.emit($event);
    }
}
SearchPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'search-panel',
                template: "<div class=\"d-flex flex-column px-2\" style=\"width: 100%;height:100%;\" #itemsContainer (mousedown)=\"onMouseDown($event)\">\r\n    <div class=\"search-tip-info d-flex flex-row align-items-center\" style=\"height: auto;line-height: normal;padding:7px\" >{{'searchbox.tips' | locale}} <i class=\"f-icon f-icon-close\" (click)=\"onClosePanel($event)\"></i></div>\r\n    <div class=\"list f-utils-fill f-datalist\" style=\"overflow: auto;\">\r\n        <ul class=\"list-group list-group-flush p-0\">\r\n            <li class=\"list-group-item list-group-item-action d-flex flex-row\" [class.active]=\"activeIndex === i\" \r\n                    *ngFor=\"let item of fields; index as i\" (click)=\"onFieldItemClick($event, item)\">\r\n                <span style=\"min-width: 60px; text-align:right\" title=\"{{item.name}} ({{ item.code }})\">{{ item.name }} </span>\r\n                <span class=\"mx-2 text-muted\" >{{'searchbox.contain' | locale}}</span>\r\n                <div class=\"f-utils-fill text-primary search-box-keyword \" [innerHTML]=\" searchKeyWord | searchwords | safe: 'html' \"></div> \r\n            </li>\r\n        </ul>\r\n    </div>\r\n</div>",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
SearchPanelComponent.ctorParameters = () => [
    { type: Injector },
    { type: ChangeDetectorRef }
];
SearchPanelComponent.propDecorators = {
    fields: [{ type: Input }],
    maxItems: [{ type: Input }],
    searchKeyWord: [{ type: Input }],
    itemClick: [{ type: Output }],
    closepanel: [{ type: Output }],
    mouseDown: [{ type: Output }],
    itemsContainer: [{ type: ViewChild, args: ['itemsContainer',] }],
    onClick: [{ type: HostListener, args: ['mousedown', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    SearchPanelComponent.prototype.fields;
    /** @type {?} */
    SearchPanelComponent.prototype.maxItems;
    /** @type {?} */
    SearchPanelComponent.prototype.searchKeyWord;
    /** @type {?} */
    SearchPanelComponent.prototype.itemClick;
    /** @type {?} */
    SearchPanelComponent.prototype.closepanel;
    /** @type {?} */
    SearchPanelComponent.prototype.mouseDown;
    /** @type {?} */
    SearchPanelComponent.prototype.itemsContainer;
    /** @type {?} */
    SearchPanelComponent.prototype.activeIndex;
    /**
     * @type {?}
     * @private
     */
    SearchPanelComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    SearchPanelComponent.prototype.cdr;
}
