// @gsp-wf/wf-task-handler/esm5/lib/entity/add-sign.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BasePayload } from "./base-payload";
/**
 * 加签模拟接口返回
 */
var /**
 * 加签模拟接口返回
 */
AddSignSimulateResponse = /** @class */ (function () {
    function AddSignSimulateResponse() {
    }
    return AddSignSimulateResponse;
}());
/**
 * 加签模拟接口返回
 */
export { AddSignSimulateResponse };
if (false) {
    /**
     * 组织
     * @type {?}
     */
    AddSignSimulateResponse.prototype.organization;
    /**
     * 多人会签方式
     * @type {?}
     */
    AddSignSimulateResponse.prototype.multiInstanceType;
    /**
     * 是否允许再加签
     * @type {?}
     */
    AddSignSimulateResponse.prototype.showSignToSign;
}
/** @enum {string} */
var MultiInstanceType = {
    /**
     * 抢占
     */
    Preempt: 'Preempt',
    /**
     * 并行
     */
    Parallel: 'Parallel',
    /**
     * 串行
     */
    Sequential: 'Sequential',
};
export { MultiInstanceType };
/**
 * 加签提交的payload
 */
var /**
 * 加签提交的payload
 */
AddSignPayload = /** @class */ (function (_super) {
    tslib_1.__extends(AddSignPayload, _super);
    function AddSignPayload() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return AddSignPayload;
}(BasePayload));
/**
 * 加签提交的payload
 */
export { AddSignPayload };
if (false) {
    /**
     * 加签方式：前加签、后加签、并加签（当前会签）、征询/协办
     * @type {?}
     */
    AddSignPayload.prototype.actionCode;
    /**
     * 多人会签方式：抢占 Preempt、并行会签 Parallel、串行会签Sequential
     * @type {?}
     */
    AddSignPayload.prototype.multiInstanceType;
    /**
     * 会签策略：全员通过allPass、一人通过onePass
     * @type {?}
     */
    AddSignPayload.prototype.counterSignRule;
    /**
     * 办理规则
     * all所有用户都办理
     * compliance符合规则往下流转
     * @type {?}
     */
    AddSignPayload.prototype.completedRule;
    /**
     * 加签人员
     * @type {?}
     */
    AddSignPayload.prototype.signAssignees;
    /**
     * 加签节点名称
     * @type {?}
     */
    AddSignPayload.prototype.activityName;
    /**
     * 允许再加签
     * @type {?}
     */
    AddSignPayload.prototype.allowSignToSign;
    /**
     * 紧急程度
     * @type {?}
     */
    AddSignPayload.prototype.priority;
}
