// @gsp-wf/wf-task-handler/esm5/lib/entity/process-cache.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 提交审批时，记下选择的流程
 */
var /**
 * 提交审批时，记下选择的流程
 */
StartProcessCache = /** @class */ (function () {
    function StartProcessCache() {
    }
    return StartProcessCache;
}());
/**
 * 提交审批时，记下选择的流程
 */
export { StartProcessCache };
if (false) {
    /** @type {?} */
    StartProcessCache.prototype.selectedProcessList;
}
/**
 * 提交审批时，记下选择的流程
 */
var /**
 * 提交审批时，记下选择的流程
 */
SelectedProcessItem = /** @class */ (function () {
    function SelectedProcessItem(processCategoryId, processDefinitionId, processDefinitionName) {
        /**
         * 记录时间
         */
        this.logTime = new Date();
        this.processCategory = processCategoryId;
        this.processDefinitionId = processDefinitionId;
        this.processDefinitionName = processDefinitionName;
        this.logTime = new Date();
    }
    return SelectedProcessItem;
}());
/**
 * 提交审批时，记下选择的流程
 */
export { SelectedProcessItem };
if (false) {
    /**
     * 流程分类id
     * @type {?}
     */
    SelectedProcessItem.prototype.processCategory;
    /**
     * 流程定义id
     * @type {?}
     */
    SelectedProcessItem.prototype.processDefinitionId;
    /**
     * 流程定义名称
     * @type {?}
     */
    SelectedProcessItem.prototype.processDefinitionName;
    /**
     * 记录时间
     * @type {?}
     */
    SelectedProcessItem.prototype.logTime;
}
