// @gsp-wf/wf-task-handler/esm5/lib/entity/task-center-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 任务中心配置
 */
var /**
 * 任务中心配置
 */
TaskCenterConfig = /** @class */ (function () {
    function TaskCenterConfig() {
    }
    return TaskCenterConfig;
}());
/**
 * 任务中心配置
 */
export { TaskCenterConfig };
if (false) {
    /** @type {?} */
    TaskCenterConfig.prototype.id;
    /** @type {?} */
    TaskCenterConfig.prototype.configCode;
    /** @type {?} */
    TaskCenterConfig.prototype.configName;
    /** @type {?} */
    TaskCenterConfig.prototype.configNote;
    /** @type {?} */
    TaskCenterConfig.prototype.defaultValue;
    /** @type {?} */
    TaskCenterConfig.prototype.currentValue;
    /** @type {?} */
    TaskCenterConfig.prototype.displayText;
    /** @type {?} */
    TaskCenterConfig.prototype.ownerId;
    /** @type {?} */
    TaskCenterConfig.prototype.ownerType;
    /** @type {?} */
    TaskCenterConfig.prototype.creator;
    /** @type {?} */
    TaskCenterConfig.prototype.createTime;
    /** @type {?} */
    TaskCenterConfig.prototype.lastModifier;
    /** @type {?} */
    TaskCenterConfig.prototype.lastModifiedTime;
}
/** @enum {string} */
var ConfigCode = {
    If_Show_Submit_Dialog: 'If_Show_Submit_Dialog',
    Info_After_Submit: 'Info_After_Submit',
    Info_After_Cancel_Submit: 'Info_After_Cancel_Submit',
    If_Use_Sign_By_Default: 'If_Use_Sign_By_Default',
    If_Show_Attachment: 'If_Show_Attachment',
    If_Show_Signature: 'If_Show_Signature',
    Default_Approval_Comment: 'Default_Approval_Comment',
    If_Store_Comment: 'If_Store_Comment',
    User_Help_Type: 'User_Help_Type',
    If_Can_Edit_Signature: 'If_Can_Edit_Signature',
};
export { ConfigCode };
/** @enum {string} */
var OwnerType = {
    User: 'user',
    Category: 'category',
    BizDefKey: 'bizdefkey',
    All: 'all',
};
export { OwnerType };
