// @farris/expression-engine/esm5/lib/expression/expression.js
import * as tslib_1 from "tslib";
import { Compiler } from "../compiler/index";
import { Evaler } from "../evaler/index";
import { DefaultFunctions } from "../function/index";
var Expression = (function () {
    function Expression(expr, context) {
        this.expr = expr;
        this.context = context;
    }
    Expression.prototype.eval = function (context) {
        context = context || this.context;
        var args = context && context.contexts && context.contexts.arguments || {};
        if (this.factory) {
            var ctx = this.buildContext(context);
            var result = this.factory(ctx.contexts)(args);
            return result;
        }
        else {
            var evaler = new Evaler();
            var factory = evaler.eval(this.expr, context);
            var result = factory.apply(void 0, tslib_1.__spread(Object.values(args)));
            return result;
        }
    };
    Expression.prototype.compile = function (context) {
        context = context || this.context;
        var compiler = new Compiler();
        this.factory = compiler.compile(this.expr, context);
        return this;
    };
    Expression.prototype.buildContext = function (context) {
        var contexts = context && context.contexts || {};
        var defaultFunction = new DefaultFunctions(context);
        contexts['DefaultFunction'] = defaultFunction;
        return context;
    };
    return Expression;
}());
export { Expression };
