// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/data-type-serializer/structure-type-ref-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/data-type-serializer/structure-type-ref-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { StructureTypeRef } from '../../extension/data-type/structure-type-ref';
var StructureTypeRefSerializer = /** @class */ (function () {
    function StructureTypeRefSerializer() {
    }
    /**
     * @param {?} type
     * @param {?} ctxt
     * @return {?}
     */
    StructureTypeRefSerializer.prototype.serialize = /**
     * @param {?} type
     * @param {?} ctxt
     * @return {?}
     */
    function (type, ctxt) {
        if (!type) {
            return null;
        }
        /** @type {?} */
        var jobj = new Object();
        /** @type {?} */
        var refType = (/** @type {?} */ (type));
        jobj['kind'] = refType.kind;
        jobj['refId'] = refType.refId;
        jobj['type'] = refType.type;
        return jobj;
    };
    /**
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    StructureTypeRefSerializer.prototype.deserialize = /**
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    function (obj, ctxt) {
        if (!obj) {
            return null;
        }
        /** @type {?} */
        var refType = new StructureTypeRef();
        refType.refId = obj['refId'];
        refType.type = (/** @type {?} */ (obj['type']));
        return refType;
    };
    return StructureTypeRefSerializer;
}());
export { StructureTypeRefSerializer };
