// @farris/ui-datagrid/esm5/lib/scrollbar/arrow/scrollbar-arrows.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
var ScrollbarArrowsComponent = /** @class */ (function () {
    function ScrollbarArrowsComponent() {
        this.reachX = 'start';
        this.reachY = 'start';
        this.scrollbarXActive = false;
        this.scrollbarYActive = false;
        this.hasXScrollbar = new EventEmitter();
        this.hasYScrollbar = new EventEmitter();
        this.scrollMoveTo = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ScrollbarArrowsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} changes
     * @return {?}
     */
    ScrollbarArrowsComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.scrollbarYActive && !changes.scrollbarYActive.isFirstChange()) {
            this.hasYScrollbar.emit(this.scrollbarYActive);
        }
    };
    /**
     * @param {?} $event
     * @param {?} dir
     * @return {?}
     */
    ScrollbarArrowsComponent.prototype.scrollTo = /**
     * @param {?} $event
     * @param {?} dir
     * @return {?}
     */
    function ($event, dir) {
        $event.stopPropagation();
        this.scrollMoveTo.emit({ dir: dir, event: $event });
    };
    ScrollbarArrowsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'scrollbar-arrows',
                    template: "<ng-container *ngIf=\"scrollbarXActive\">\r\n    <div class=\"scrollbar-arrow scrollbar-arrow-left\" (click)=\"scrollTo($event, 'left')\" [class.scrollbar-arrow-disable]=\"reachX === 'start'\">\r\n        <span class=\"f-icon f-icon-arrow-60-left\"></span>\r\n    </div>\r\n    <div class=\"scrollbar-arrow scrollbar-arrow-right\" [class.scrollbar-y]=\"scrollbarYActive\" (click)=\"scrollTo($event, 'right')\"  [class.scrollbar-arrow-disable]=\"reachX === 'end'\">\r\n        <span class=\"f-icon f-icon-arrow-60-right\"></span>\r\n    </div>\r\n</ng-container>\r\n\r\n\r\n\r\n<ng-container *ngIf=\"scrollbarYActive\">\r\n    <div class=\"scrollbar-arrow scrollbar-arrow-top\" (click)=\"scrollTo($event, 'top')\"  [class.scrollbar-arrow-disable]=\"reachY === 'start'\">\r\n        <span class=\"f-icon f-icon-arrow-60-up\"></span>\r\n    </div>\r\n    <div class=\"scrollbar-arrow scrollbar-arrow-bottom\" [class.scrollbar-x]=\"scrollbarXActive\" (click)=\"scrollTo($event, 'bottom')\"  [class.scrollbar-arrow-disable]=\"reachY === 'end'\">\r\n        <span class=\"f-icon f-icon-arrow-60-down\"></span>\r\n    </div>\r\n</ng-container>\r\n\r\n<div class=\"scrollbar-arrow-block\" *ngIf=\"scrollbarYActive && scrollbarXActive\"></div>\r\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".scrollbar-allways-show .ps.ps--active-x>.ps__rail-x{margin-right:0!important}.scrollbar-allways-show .ps.ps--active-y>.ps__rail-y{margin-bottom:0!important}.scrollbar-allways-show .ps.ps--active-x.ps--active-y>.ps__rail-x{margin-right:18px!important}.scrollbar-allways-show .ps.ps--active-x.ps--active-y>.ps__rail-y{margin-bottom:18px!important}.scrollbar-allways-show.scroolbar-arrows .ps.ps--active-x>.ps__rail-x{margin-right:18px!important;margin-left:18px!important}.scrollbar-allways-show.scroolbar-arrows .ps.ps--active-y>.ps__rail-y{margin-top:18px!important;margin-bottom:18px!important}.scrollbar-allways-show.scroolbar-arrows .ps.ps--active-x.ps--active-y>.ps__rail-x{margin-right:36px!important}.scrollbar-allways-show.scroolbar-arrows .ps.ps--active-x.ps--active-y>.ps__rail-y{margin-bottom:36px!important}.scroolbar-arrows .scrollbar-arrow{position:absolute;width:18px;height:18px;z-index:1000;background:#f5f6f9;color:#505050}.scroolbar-arrows .scrollbar-arrow:hover:not(.scrollbar-arrow-disable){background:#d2d8e9}.scroolbar-arrows .scrollbar-arrow:active:not(.scrollbar-arrow-disable){background:#b2b2b2;color:#fff}.scroolbar-arrows .scrollbar-arrow-left{bottom:0;border-top:1px solid #dee4ed}.scroolbar-arrows .scrollbar-arrow-right{right:0;bottom:0;border-top:1px solid #dee4ed}.scroolbar-arrows .scrollbar-arrow-bottom,.scroolbar-arrows .scrollbar-arrow-left,.scroolbar-arrows .scrollbar-arrow-right,.scroolbar-arrows .scrollbar-arrow-top{display:flex;align-items:center;justify-content:center}.scroolbar-arrows .scrollbar-arrow-left .f-icon,.scroolbar-arrows .scrollbar-arrow-right .f-icon{position:relative;top:-1px;font-size:12px}.scroolbar-arrows .scrollbar-arrow-bottom .f-icon,.scroolbar-arrows .scrollbar-arrow-top .f-icon{font-size:12px}.scroolbar-arrows .scrollbar-arrow-right.scrollbar-y{right:18px}.scroolbar-arrows .scrollbar-arrow-top{top:0;right:0;border-left:1px solid #dee4ed}.scroolbar-arrows .scrollbar-arrow-bottom{bottom:0;right:0;border-left:1px solid #dee4ed}.scroolbar-arrows .scrollbar-arrow-bottom.scrollbar-x{bottom:18px}.scroolbar-arrows .scrollbar-arrow-block{width:18px;height:18px;position:absolute;right:0;bottom:0;background:#fff;z-index:1000}.scroolbar-arrows .scrollbar-arrow-disable{color:#dadada}"]
                }] }
    ];
    /** @nocollapse */
    ScrollbarArrowsComponent.ctorParameters = function () { return []; };
    ScrollbarArrowsComponent.propDecorators = {
        reachX: [{ type: Input }],
        reachY: [{ type: Input }],
        scrollbarXActive: [{ type: Input }],
        scrollbarYActive: [{ type: Input }],
        hasXScrollbar: [{ type: Output }],
        hasYScrollbar: [{ type: Output }],
        scrollMoveTo: [{ type: Output }]
    };
    return ScrollbarArrowsComponent;
}());
export { ScrollbarArrowsComponent };
if (false) {
    /** @type {?} */
    ScrollbarArrowsComponent.prototype.reachX;
    /** @type {?} */
    ScrollbarArrowsComponent.prototype.reachY;
    /** @type {?} */
    ScrollbarArrowsComponent.prototype.scrollbarXActive;
    /** @type {?} */
    ScrollbarArrowsComponent.prototype.scrollbarYActive;
    /** @type {?} */
    ScrollbarArrowsComponent.prototype.hasXScrollbar;
    /** @type {?} */
    ScrollbarArrowsComponent.prototype.hasYScrollbar;
    /** @type {?} */
    ScrollbarArrowsComponent.prototype.scrollMoveTo;
}
