// @farris/ui-batch-edit-dialog/esm2015/lib/batch-edit-dialog.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ComponentFactoryResolver, Injector, EventEmitter } from '@angular/core';
import { BatchEditDialogComponent } from './batch-edit-dialog.component';
import { BsModalService } from '@farris/ui-modal';
import { LocaleService } from '@farris/ui-locale';
import { of, Subject } from 'rxjs';
import { ResultConfirmComponent } from './result-confirm/result-confirm.component';
// tslint:disable: max-line-length
export class BatchEditDialogService {
    /**
     * @param {?} injector
     * @param {?} modalService
     * @param {?} componentFactoryResolver
     * @param {?} localeService
     */
    constructor(injector, modalService, componentFactoryResolver, localeService) {
        this.injector = injector;
        this.modalService = modalService;
        this.componentFactoryResolver = componentFactoryResolver;
        this.localeService = localeService;
        this.dialogRef = null;
        this.componentRef = null;
        // 用户对输入进行确认，此时可以更新bindingData数据
        this.onUserConfirmed = new EventEmitter();
    }
    /**
     * \@description 控件参数格式
     * # 控件参数格式
     * ## textbox
     * ```json
     * {"title":"姓名","field":"name","editor":{"controlType":EditorTypes.TEXTBOX}}
     * ```
     * ## dropdown
     * ```json
     * {"title":"姓名","field":"name","editor":{"controlType":EditorTypes.COMBOLIST,"idField":"id","textField":"text","items":[],"multiSelect":false}}
     * ```
     * ## lookup
     * ```json
     * {"title":"姓名","field":"name","editor":{"controlType":EditorTypes.LOOKUP,"uri":"","idField":"id","textField":"text","valueField":"id","displayType":"LIST","mapFields":{},"singleSelect":true,"dialogTitle":"","hiddenInputName":"","items":[],"multiSelect":false}}
     * ```
     * ## 数值输入框
     * ```json
     * {"title":"数值","field":"name","editor":{"controlType":EditorTypes.NUMBER,"precision":2,"step":1,"min":1,"max":10}}
     * ```
     * @param {?} columns
     * @param {?=} options
     * @return {?}
     */
    batchEdit(columns, options) {
        /** @type {?} */
        const titleText = this.localeService.getValue('batchEditDialog.title');
        /** @type {?} */
        const okText = this.localeService.getValue('batchEditDialog.okText');
        /** @type {?} */
        const cancelText = this.localeService.getValue('batchEditDialog.cancelText');
        const { width = 792, height = 580, title = titleText, rows = 1 } = options || {};
        /** @type {?} */
        const componentFactory = this.componentFactoryResolver.resolveComponentFactory(BatchEditDialogComponent);
        this.componentRef = componentFactory.create(this.injector);
        this.componentRef.instance.fields = this.getFields(columns);
        this.componentRef.instance.eidtors = this.getEditors(columns);
        this.componentRef.instance.title = title;
        this.componentRef.instance.rows = rows;
        this.dialogRef = this.modalService.show(this.componentRef, {
            title,
            width,
            height,
            buttons: [
                {
                    text: cancelText, cls: 'btn btn-secondary', handle: (/**
                     * @return {?}
                     */
                    () => {
                        this.dialogRef.close();
                    })
                },
                {
                    text: okText, cls: 'btn btn-primary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        this.userConfired();
                    })
                }
            ],
            showHeader: false,
            showButtons: false,
        });
        this.componentRef.instance.confirmed.subscribe((/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            this.canConfirm(options).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            (result) => {
                if (result === true) {
                    this.userConfired(options);
                    this.dialogRef.close();
                }
            }));
        }));
        this.componentRef.instance.canceled.subscribe((/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            this.dialogRef.close();
        }));
        this.componentRef.instance.dialogRef = this.dialogRef;
        this.dialogRef.dialog.instance.draggbar.handle = this.componentRef.instance.header.nativeElement;
        return this.dialogRef;
    }
    /**
     * 用户确认修改
     * @param {?=} options
     * @return {?}
     */
    userConfired(options) {
        /** @type {?} */
        const inputs = this.componentRef && this.componentRef.instance && this.componentRef.instance.inputs || [];
        // 针对checkbox做特殊处理
        if (inputs && inputs.length > 0) {
            inputs.forEach((/**
             * @param {?} input
             * @return {?}
             */
            input => {
                const { controlType, dataType } = input;
                if (controlType === 'checkbox' && dataType === 'boolean') {
                    input.value = input.value ? true : false;
                }
            }));
        }
        if (options && options.onConfirm && typeof options.onConfirm === 'function') {
            options.onConfirm(inputs);
        }
        this.onUserConfirmed.next(inputs);
    }
    /**
     * 从dom中获取grid列
     * @private
     * @param {?} columns dom
     * @return {?}
     */
    getFields(columns) {
        /** @type {?} */
        const fields = [];
        if (!columns || columns.length < 1) {
            return fields;
        }
        columns.forEach((/**
         * @param {?} collect
         * @return {?}
         */
        (collect) => {
            // collect为数组
            if (collect && collect.length > 0) {
                collect.forEach((/**
                 * @param {?} column
                 * @return {?}
                 */
                (column) => {
                    const { field = null, title = '', dataType = null, enableBatchEdit = false } = column || {};
                    if (field && enableBatchEdit) {
                        fields.push({ id: field, label: title, dataType });
                    }
                }));
            }
        }));
        return fields;
    }
    /**
     * 从columns中获取编辑器
     * @private
     * @param {?} columns dom
     * @return {?}
     */
    getEditors(columns) {
        /** @type {?} */
        const editors = [];
        if (!columns || columns.length < 1) {
            return editors;
        }
        columns.forEach((/**
         * @param {?} collect
         * @return {?}
         */
        (collect) => {
            // collect为数组
            if (collect && collect.length > 0) {
                collect.forEach((/**
                 * @param {?} column
                 * @return {?}
                 */
                (column) => {
                    const { field = null, editor = {}, enableBatchEdit = false } = column || {};
                    if (editor && field && enableBatchEdit) {
                        editor.controlType = editor.type;
                        editors.push({ field, editor });
                    }
                }));
            }
        }));
        return editors;
    }
    /**
     * @private
     * @param {?=} options
     * @return {?}
     */
    canConfirm(options) {
        /** @type {?} */
        const subject = new Subject();
        /** @type {?} */
        let showNotify = options && options.showNotify;
        const { rows = 1 } = options || {};
        /** @type {?} */
        const valueChanged = this.isValueChanged();
        if (!valueChanged) {
            return of(true);
        }
        if (showNotify === undefined) {
            showNotify = true;
        }
        if (!showNotify) {
            return of(true);
        }
        /** @type {?} */
        let dialogRef = null;
        try {
            /** @type {?} */
            const componentFactory = this.componentFactoryResolver.resolveComponentFactory(ResultConfirmComponent);
            /** @type {?} */
            const componentRef = componentFactory.create(this.injector);
            /** @type {?} */
            const shouldShowNotify = componentRef.instance.showNotify || false;
            componentRef.instance.rows = rows;
            if (shouldShowNotify) {
                dialogRef = this.modalService.show(componentRef, {
                    title: '',
                    width: 500,
                    height: 240,
                    showHeader: false,
                    showButtons: false,
                    buttons: []
                });
                // 确认
                componentRef.instance.confirmed.subscribe((/**
                 * @param {?} e
                 * @return {?}
                 */
                (e) => {
                    if (dialogRef) {
                        dialogRef.close();
                    }
                    subject.next(true);
                }));
                // 取消
                componentRef.instance.canceled.subscribe((/**
                 * @param {?} e
                 * @return {?}
                 */
                (e) => {
                    if (dialogRef) {
                        dialogRef.close();
                    }
                    subject.next(false);
                }));
            }
            else {
                return of(true);
            }
        }
        catch (e) {
            return of(true);
        }
        return subject;
    }
    /**
     * @return {?}
     */
    isValueChanged() {
        /** @type {?} */
        const inputs = this.componentRef && this.componentRef.instance && this.componentRef.instance.inputs || [];
        if (!inputs || inputs.length < 1) {
            return false;
        }
        /** @type {?} */
        const unChanged = inputs.every((/**
         * @param {?} item
         * @return {?}
         */
        (item) => !item.field));
        return !unChanged;
    }
}
BatchEditDialogService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
BatchEditDialogService.ctorParameters = () => [
    { type: Injector },
    { type: BsModalService },
    { type: ComponentFactoryResolver },
    { type: LocaleService }
];
if (false) {
    /** @type {?} */
    BatchEditDialogService.prototype.dialogRef;
    /**
     * @type {?}
     * @private
     */
    BatchEditDialogService.prototype.componentRef;
    /** @type {?} */
    BatchEditDialogService.prototype.onUserConfirmed;
    /** @type {?} */
    BatchEditDialogService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    BatchEditDialogService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    BatchEditDialogService.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    BatchEditDialogService.prototype.localeService;
}
