// @farris/ui-batch-edit-dialog/esm2015/lib/lookup.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { LookupComponent } from '@farris/ui-lookup';
import { of } from 'rxjs';
export class LookupDirective {
    constructor() {
        this.onSelectRows = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.lookup) {
            this.lookup.dictPicked = (/**
             * @param {?} rows
             * @return {?}
             */
            (rows) => {
                if (this.onSelectRows) {
                    this.onSelectRows.emit({ field: this.field, rows: [].concat(rows) });
                }
                /** @type {?} */
                const result = {
                    closeDialog: true
                };
                return of(result);
            });
        }
    }
}
LookupDirective.decorators = [
    { type: Directive, args: [{
                selector: '[lookup-extra]'
            },] }
];
/** @nocollapse */
LookupDirective.ctorParameters = () => [];
LookupDirective.propDecorators = {
    field: [{ type: Input, args: ['lookup-extra',] }],
    lookup: [{ type: Input, args: ['ins',] }],
    onSelectRows: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    LookupDirective.prototype.field;
    /** @type {?} */
    LookupDirective.prototype.lookup;
    /** @type {?} */
    LookupDirective.prototype.onSelectRows;
}
