// @farris/external-integration/fesm2015/farris-external-integration.js
import { DatagridBaseEditorDirective } from '@farris/ui-datagrid-editors';
import { of } from 'rxjs';
import { Injectable, Output, EventEmitter, Component, ComponentFactoryResolver, ElementRef, forwardRef, Injector, Input, Optional, ViewChild, ViewContainerRef, Renderer2, Directive, NgModule } from '@angular/core';
import { VariableParseService, ViewModel } from '@farris/devkit';
import { CommonModule } from '@angular/common';
import { DialogService, FarrisDialogModule } from '@farris/ui-dialog';
import { BsModalService, ModalModule } from '@farris/ui-modal';
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { NgControl, NG_VALUE_ACCESSOR, FormsModule, ReactiveFormsModule } from '@angular/forms';
import { LocaleModule } from '@farris/ui-locale';
import { PerfectScrollbarModule } from '@farris/ui-perfect-scrollbar';
import { TreeModule } from '@farris/ui-tree';
import { FarrisTabsModule } from '@farris/ui-tabs';
import { RuntimeStateService, FarrisCommonModule } from '@farris/ui-common';
import { FarrisFormsModule } from '@farris/ui-forms';
import { TextModule } from '@farris/ui-text';
import { MessagerModule } from '@farris/ui-messager';
import { NotifyModule } from '@farris/ui-notify';
import { LayoutModule } from '@farris/ui-layout';
import { GRID_EDITORS, DatagridComponent, DatagridModule } from '@farris/ui-datagrid';
import { InputGroupModule } from '@farris/ui-input-group';
import { FormMessageModule } from '@farris/ui-forms/validation-message';
import { ListNavModule } from '@farris/ui-list-nav';
import { ListViewModule } from '@farris/ui-list-view';
import { RouterModule } from '@angular/router';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/component/external-integration/external-integration-service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ExternalIntegrationService {
    /**
     * @param {?} http
     * @param {?} variableParseService
     * @param {?} vm
     */
    constructor(http, variableParseService, vm) {
        this.http = http;
        this.variableParseService = variableParseService;
        this.vm = vm;
        this.sapData = new EventEmitter();
        this.sapUrl = "/api/runtime/icc/v1.0/externalServiceOpenApi/invokeService";
    }
    /**
     * 调用外部服务
     * @param {?} externalParams
     * @param {?} serviceCode
     * @return {?}
     */
    queryExtIntegrationService(externalParams, serviceCode) {
        /** @type {?} */
        let req = this.assembleSapRequest(externalParams, serviceCode);
        this.http.post(this.sapUrl, req).subscribe((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            /** @type {?} */
            let respObj = Object.create(response);
            if (!respObj || !respObj.success) {
                return;
            }
            /** @type {?} */
            let resp = Object.create(respObj);
            /** @type {?} */
            let sapData = resp.data;
            this.sapData.emit(sapData);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.log(error);
            this.sapData.emit({});
        }));
    }
    /**
     * 组装sap api入参
     * @private
     * @param {?} externalParams
     * @param {?} serviceCode
     * @return {?}
     */
    assembleSapRequest(externalParams, serviceCode) {
        /** @type {?} */
        const params = JSON.parse(JSON.stringify(externalParams));
        /** @type {?} */
        let requestParam = {
            code: serviceCode,
            params: this.variableParseService.parse(params, this.vm.frameContext),
        };
        return requestParam;
    }
}
ExternalIntegrationService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ExternalIntegrationService.ctorParameters = () => [
    { type: HttpClient },
    { type: VariableParseService },
    { type: ViewModel }
];
ExternalIntegrationService.propDecorators = {
    sapData: [{ type: Output }]
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/component/external-integration/external-integration.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ExternalIntegrationComponent {
    /**
     * @param {?} vm
     * @param {?} el
     * @param {?} extIntgrtnSrvc
     * @param {?} variableParseService
     * @param {?} resolver
     * @param {?} inj
     */
    constructor(vm, el, extIntgrtnSrvc, variableParseService, resolver, inj) {
        this.vm = vm;
        this.el = el;
        this.extIntgrtnSrvc = extIntgrtnSrvc;
        this.variableParseService = variableParseService;
        this.resolver = resolver;
        this.inj = inj;
        /**
         * 模态框宽
         */
        this.width = 960;
        /**
         * 模态框高
         */
        this.height = 577;
        this.fieldText = "";
        this.placeholder = "请选择";
        // 是否只读
        this.readonly = false;
        /**
         * 内容中留白边距
         */
        this.containerMargin = { top: 0, bottom: 5, left: 10, right: 10 };
        /**
         * 指令回填sap数据
         */
        this.selectSapData = new EventEmitter();
        this.afterOnblur = new EventEmitter();
        /**
         * 多表数据集合
         */
        this.tableData = [];
        /**
         * 模态框导航栏当前选中项
         */
        this.currentLink = {};
        /**
         * 模态框确定按钮是否可用
         */
        this.ensureFlag = true;
        this.containerStyle = {
            marginLeft: this.containerMargin.left + "px",
            marginRight: this.containerMargin.right + "px",
            marginTop: this.containerMargin.top + "px",
            marginBottom: this.containerMargin.bottom + "px"
        };
        this.beforeClose = (/**
         * @return {?}
         */
        () => {
            this.tableData = [];
            this.ensureFlag = true;
            this.sendSapData();
            return of(true);
        });
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) { }
    /**
     * @return {?}
     */
    fieldTextChange() {
        /** @type {?} */
        const pathArr = this.getBindingPathArray();
        this.vm.bindingData.setValue(pathArr.concat(this.ngControl.name), this.fieldText, true, true);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() { }
    /**
     * @return {?}
     */
    ngOnDestroy() { }
    /**
     * @param {?} obj
     * @return {?}
     */
    writeValue(obj) {
        this.fieldText = obj;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) { }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) { }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) { }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.ngControl = this.inj.get(NgControl);
        // 卡片绑定onblur事件，表格再grid-external-integration中绑定onblur事件
        if (this.mappingType != "grid") {
            this.input.nativeElement.onblur = (/**
             * @return {?}
             */
            () => {
                this.callSapService();
            });
        }
        this.extIntgrtnSrvc.sapData.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            this.sapDataHandle(data);
        }));
    }
    /**
     * 打开模态框
     * @return {?}
     */
    showDialog() {
        this.farrisListView.listClick.emit({
            data: [this.tableData[0]],
            index: 0,
            checkChangeEvent: false
        });
        this.farrisListView.clickItem = this.tableData[0];
        this.dialog.show();
        return false;
    }
    /**
     * 调用sap外部服务，出参在ngOninit监听
     * @return {?}
     */
    callSapService() {
        this.fieldTextChange();
        this.extIntgrtnSrvc.queryExtIntegrationService(this.externalParams, this.serviceCode);
    }
    /**
     * 处理sap出参
     * @param {?} data sap出参
     * @return {?}
     */
    sapDataHandle(data) {
        if (!data) {
            return;
        }
        /** @type {?} */
        let tableNameList = Object.keys(data);
        /** @type {?} */
        let mapTableList = this.getTableListFromMapFields();
        // 取交集
        tableNameList = tableNameList.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            return mapTableList.indexOf(item) != -1;
        }));
        // 多表场景
        if (tableNameList.length > 1) {
            // 校验是否弹出模态框
            /** @type {?} */
            let openDialogFlag = this.checkMultiTableData(data, tableNameList);
            if (openDialogFlag == "open") {
                // 组装模态框数据
                this.assembleTableData(data);
                this.showDialog();
            }
            else if (openDialogFlag == "backfill") {
                // 多表，每个表只有一条数据
                /** @type {?} */
                let backfillData = {};
                tableNameList.forEach((/**
                 * @param {?} tableName
                 * @return {?}
                 */
                tableName => {
                    Object.assign(backfillData, this.assembleSapData(tableName, data[tableName][0]));
                }));
                this.sendSapData(backfillData);
            }
            else {
                // 多表无数据
                this.sendSapData();
            }
        }
        else if (tableNameList.length == 1) {
            /** @type {?} */
            let tableName = tableNameList[0];
            /** @type {?} */
            let sapDataLength = (data[tableName] || []).length;
            if (sapDataLength > 0) {
                if (sapDataLength == 1) {
                    // 一个表只有一条数据
                    /** @type {?} */
                    let sapData = this.assembleSapData(tableName, data[tableName][0]);
                    this.sendSapData(sapData);
                }
                else if (sapDataLength > 1) {
                    // 一个表有多条数据
                    this.assembleTableData(data);
                    this.showDialog();
                    return;
                }
            }
        }
        else {
            // 处理查询异常，editor没有endPending
            this.sendSapData();
        }
    }
    /**
     * @param {?} data
     * @param {?} tableNameList
     * @return {?}
     */
    checkMultiTableData(data, tableNameList) {
        /** @type {?} */
        let openDialogFlag = "";
        for (let index = 0; index < tableNameList.length; index++) {
            /** @type {?} */
            let tableName = tableNameList[index];
            if (data[tableName].length > 1) {
                openDialogFlag = "open";
                break;
            }
            else if (data[tableName].length == 0 && openDialogFlag != "backfill") {
                openDialogFlag = "none";
                continue;
            }
            openDialogFlag = "backfill";
        }
        return openDialogFlag;
    }
    /**
     * 组装模态框数据
     * @param {?} data sap出参
     * @return {?}
     */
    assembleTableData(data) {
        /** @type {?} */
        let tableNameList = Object.keys(data);
        /** @type {?} */
        let mapTableList = this.getTableListFromMapFields();
        // 取交集
        mapTableList = mapTableList.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            return tableNameList.indexOf(item) != -1;
        }));
        if (mapTableList.length > 0) {
            // 遍历sap映射表，组装表及对应数据
            mapTableList.forEach((/**
             * @param {?} mapTable
             * @return {?}
             */
            mapTable => {
                // 组装模态框中单表数据
                /** @type {?} */
                let tableInfo = this.assembleTableInfo(mapTable, data);
                if (tableInfo &&
                    Object.keys(tableInfo).length > 0 &&
                    (tableInfo.data || []).length > 0) {
                    this.tableData.push(tableInfo);
                }
            }));
        }
    }
    /**
     * @return {?}
     */
    getTableListFromMapFields() {
        // 获取需要回填的表
        /** @type {?} */
        let mapKeyList = Object.keys(this.myField);
        // 定义sap映射表集合
        /** @type {?} */
        let mapTableList = [];
        (mapKeyList || []).forEach((/**
         * @param {?} mapKey
         * @return {?}
         */
        mapKey => {
            // 在映射关系中获取sap出参中的表集合
            mapTableList.push(mapKey.split(".")[0]);
        }));
        // 去重
        mapTableList = Array.from(new Set(mapTableList));
        return mapTableList;
    }
    /**
     * 组装模态框中单表数据
     * @param {?} mapTable 表名
     * @param {?} data sap出参
     * @return {?} 模态框单表数据对象
     */
    assembleTableInfo(mapTable, data) {
        // 模态框单表数据对象
        /** @type {?} */
        let tableInfo = {};
        // 表头集合
        /** @type {?} */
        let tableColumnList = [];
        // 遍历sap出参表信息，获取表头中文、表名
        (this.extTableSchemas || []).forEach((/**
         * @param {?} tableSchema
         * @return {?}
         */
        tableSchema => {
            if (tableSchema.code == mapTable) {
                tableInfo.name = tableSchema.name;
                // 模拟主键，保证farris-grid组件【单选】功能正常
                /** @type {?} */
                let idColumn = {};
                idColumn.field = "id";
                idColumn.title = "序号";
                tableColumnList.push(idColumn);
                // 组装表头
                tableSchema.columns.forEach((/**
                 * @param {?} column
                 * @return {?}
                 */
                column => {
                    /** @type {?} */
                    let tableColumn = {};
                    tableColumn.field = column.code;
                    tableColumn.title = column.name;
                    tableColumnList.push(tableColumn);
                }));
                tableInfo.columns = tableColumnList;
                tableInfo.code = mapTable;
                // 给表数据增加模拟主键
                (data[mapTable] || []).forEach((/**
                 * @param {?} item
                 * @param {?} index
                 * @return {?}
                 */
                (item, index) => {
                    item["id"] = index + 1;
                }));
                tableInfo.data = data[mapTable];
            }
        }));
        return tableInfo;
    }
    /**
     * 导航栏点击事件
     * @param {?} event 导航栏信息
     * @return {?}
     */
    listClick(event) {
        /** @type {?} */
        let clickData = event.data;
        if (event.data.disable) {
            return false;
        }
        else {
            if (typeof this.currentLink["data"] !== "undefined") {
                if (clickData[0].code === this.currentLink["data"][0].code) {
                    return false;
                }
                else {
                    this.currentLink = event;
                }
            }
            else {
                this.currentLink = event;
            }
        }
    }
    /**
     * 模态框，确定，取消按钮
     * @param {?} flag ensure:确定按钮，cancel：取消按钮
     * @return {?}
     */
    confirm(flag) {
        // 点击确定按钮
        if (flag == "ensure") {
            /** @type {?} */
            let sapData = {};
            // 遍历模态框表数据，组装回填数据
            for (let index = 0; index < this.tableData.length; index++) {
                /** @type {?} */
                let tableInfo = this.tableData[index];
                Object.assign(sapData, this.assembleSapData(tableInfo.code, tableInfo.selectedData.data));
            }
            this.sendSapData(sapData);
        }
        // 清空模态框数据
        this.tableData = [];
        // 关闭模态框
        this.dialog.close();
    }
    /**
     * 选中一行事件
     * @param {?} selectedRow 选中行数据
     * @param {?} tableName 当前表格对应表名
     * @return {?}
     */
    selectRowChange(selectedRow, tableName) {
        this.tableData.forEach((/**
         * @param {?} tableInfo
         * @return {?}
         */
        tableInfo => {
            if (tableInfo.code == tableName) {
                tableInfo.selectedData = selectedRow;
            }
        }));
        this.setEnsureButtonFlag();
    }
    /**
     * 取消一行事件
     * @param {?} selectedRow 被取消行数据
     * @param {?} tableName 当前表格对应表名
     * @return {?}
     */
    unSelect(selectedRow, tableName) {
        this.tableData.forEach((/**
         * @param {?} tableInfo
         * @return {?}
         */
        tableInfo => {
            if (tableInfo.code == tableName) {
                tableInfo.selectedData = null;
                tableInfo.selectedValue = "";
            }
        }));
        selectedRow = null;
        this.setEnsureButtonFlag();
    }
    /**
     * 设置模态框确定按钮是否可用
     * @return {?}
     */
    setEnsureButtonFlag() {
        /** @type {?} */
        let flag = false;
        for (let index = 0; index < this.tableData.length; index++) {
            /** @type {?} */
            let tableInfo = this.tableData[index];
            if (!tableInfo.selectedData) {
                flag = true;
                break;
            }
        }
        this.ensureFlag = flag;
    }
    /**
     * 组装sap回填数据
     * @param {?} tableName 表名
     * @param {?} tableData 表数据
     * @return {?} (表名.表字段名:表数据):Object
     */
    assembleSapData(tableName, tableData) {
        if (!tableName || !tableData) {
            return;
        }
        // sap回填数据
        /** @type {?} */
        let sapData = {};
        // 表字段名集合
        /** @type {?} */
        let tableDataItemKeys = Object.keys(tableData);
        for (let index = 0; index < tableDataItemKeys.length; index++) {
            /** @type {?} */
            let itemKey = tableDataItemKeys[index];
            // sap回填数据key= 表名.表字段名
            /** @type {?} */
            let key = tableName + "." + itemKey;
            sapData[key] = tableData[itemKey];
        }
        return sapData;
    }
    /**
     * 回填sap数据
     * @private
     * @param {?=} sapData 将要回填的sap数据
     * @return {?}
     */
    sendSapData(sapData) {
        if (sapData) {
            this.selectSapData.emit(sapData);
        }
        // 子表关闭editor状态
        if (this.mappingType == "grid") {
            this.afterOnblur.emit();
        }
    }
    /**
     * @private
     * @return {?}
     */
    getBindingPathArray() {
        /** @type {?} */
        const path = this.vm.bindingPath;
        if (path) {
            return path.split("/").filter((/**
             * @param {?} n
             * @return {?}
             */
            n => n !== ""));
        }
        return [];
    }
}
ExternalIntegrationComponent.decorators = [
    { type: Component, args: [{
                selector: "web-external-integration",
                template: "<input [(ngModel)]=\"fieldText\" #input  class=\"form-control\" input-end-edit [readonly]=\"false\"\r\n  tabindex=\"0\" maxlength=\"36\" />\r\n\r\n<farris-dialog #dialog [title]=\"'\u9009\u62E9\u6570\u636E'\" [beforeClose]=\"beforeClose\"  [width]=\"1000\" [height]=\"834\" [showButtons]=\"true\" [showMaxButton]=\"true\"\r\n  [showCloseButton]=\"true\" [enableScroll]=\"false\" [dialogHeaderHeight]=\"50\" [buttons]=\"defaultButtonRef\">\r\n  <div [ngStyle]=\"containerStyle\" style=\"height: 100%;\">\r\n    <div class=\"listnav-example-wrapper\">\r\n      <div class=\"example-side-nav\">\r\n        <farris-list-nav  [listNavWidth]=\"240\" (listClick)=\"listClick($event)\">\r\n          <ng-template listNavContent>\r\n            <farris-list-view #farrisListView [showEmpty]=\"true\" [data]=\"tableData\" [activeIndex]=\"3\" listidName=\"code\" (listClick)=\"listClick($event)\">\r\n              <ng-template #navView listTemplate let-item=\"item\" let-selected=\"selectedItem\">\r\n                <div class=\"f-template-listnav-row\">\r\n                  <!-- routerLinkActive=\"active\" -->\r\n                  <a class=\"list-nav-link\" [ngStyle]=\"{'pointer-events': item.disable ?'none' : '','width':240 +'px'}\"\r\n                     [title]=\"item.name\">\r\n                    <span class=\"nav-item-name\">\r\n                      {{ item.name }}\r\n                    </span>\r\n                  </a>\r\n                </div>\r\n              </ng-template>\r\n            </farris-list-view>\r\n          </ng-template>\r\n\r\n          <!-- <ng-template listNavFooter>\r\n                \u5E95\u90E8\u90E8\u5206\r\n            </ng-template> -->\r\n\r\n        </farris-list-nav>\r\n      </div>\r\n      <div class=\"example-content\" >\r\n        <ng-container #girdView *ngFor=\"let item of tableData;let i = index\">\r\n          <div style=\"height:100%\" [hidden]=\"! (currentLink && item.code == ((currentLink['data'] || [])[0] || {})['code'])\" >\r\n            <farris-datagrid \r\n              [autoFitColumns]=\"true\" \r\n              [columns]=\"item.columns\" \r\n              [data]=\"item.data\" \r\n              [fitColumns]=\"true\"\r\n              [fit]=\"true\" \r\n              [showBorder]=\"true\" \r\n              [(selectValue)]=\"item.selectedValue\"\r\n              [keepSelect]=\"false\"\r\n              [pagination]=\"false\"\r\n              (selectChanged)=\"selectRowChange($event, item.code)\"\r\n              (unSelect)=\"unSelect(null, item.code)\">\r\n            </farris-datagrid>\r\n          </div>\r\n        </ng-container>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n\r\n<ng-template #defaultButtonRef>\r\n  <button #okbtn type=\"button\" [disabled]=\"ensureFlag\" (click)=\"confirm('ensure')\"  class=\"btn btn-primary btn-lg\">\r\n    {{ 'lookup.okText' | locale: '\u786E\u5B9A' }}\r\n  </button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-lg\" (click)=\"confirm('cancel')\" [disabled]=\"false\">\r\n    {{ 'lookup.cancelText' | locale: '\u53D6\u6D88' }}\r\n  </button>\r\n</ng-template>",
                providers: [
                    ExternalIntegrationService,
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => ExternalIntegrationComponent)),
                        multi: true
                    }
                ],
                styles: [".listnav-example-wrapper{background-color:#e7ebef;display:flex;height:100%}.listnav-example-wrapper .example-side-nav{flex:0 0 auto}.listnav-example-wrapper .example-content{flex:1 1 0;width:726px;overflow-y:auto;padding:10px 10px 10px 4px}.listnav-example-wrapper .example-content .listnav-content{height:100%;width:100%;background-color:#fff}.listnav-example-wrapper .example-content.example-content-right{padding:10px 4px 10px 10px}"]
            }] }
];
/** @nocollapse */
ExternalIntegrationComponent.ctorParameters = () => [
    { type: ViewModel, decorators: [{ type: Optional }] },
    { type: ElementRef },
    { type: ExternalIntegrationService },
    { type: VariableParseService },
    { type: ComponentFactoryResolver },
    { type: Injector }
];
ExternalIntegrationComponent.propDecorators = {
    width: [{ type: Input }],
    height: [{ type: Input }],
    serviceCode: [{ type: Input, args: ["serviceCode",] }],
    externalParams: [{ type: Input, args: ["externalParams",] }],
    mappingType: [{ type: Input, args: ["mappingType",] }],
    myField: [{ type: Input, args: ["mapFields",] }],
    fieldText: [{ type: Input }],
    extTableSchemas: [{ type: Input, args: ["extTableSchemas",] }],
    placeholder: [{ type: Input }],
    readonly: [{ type: Input }],
    beforeSapSrvc: [{ type: Input }],
    customData: [{ type: Input }],
    selectSapData: [{ type: Output }],
    afterOnblur: [{ type: Output }],
    dialog: [{ type: ViewChild, args: ["dialog",] }],
    input: [{ type: ViewChild, args: ["input",] }],
    girdView: [{ type: ViewChild, args: ["girdView", { read: ViewContainerRef },] }],
    farrisListView: [{ type: ViewChild, args: ["farrisListView",] }]
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/component/external-integration/grid-external-integration.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const ExtIntgrtnDefautOption = {
    readonly: false,
    // 是否只读
    placeholder: "请选择",
    viewType: "tag"
};
class DatagridExtIntgrtnComponent extends DatagridBaseEditorDirective {
    /**
     * @param {?} render
     * @param {?} el
     * @param {?} rts
     * @param {?} injector
     */
    constructor(render, el, rts, injector) {
        super(render, el, injector);
        this.rts = rts;
        this.stopPropagation = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.inputElement = this.instance.input.nativeElement;
        this.options = Object.assign({}, ExtIntgrtnDefautOption, this.options);
        this.column.editor.options.onBlur = (/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            e.editorRef.startPending();
            e.editorRef.formControl.setValue(this.instance.fieldText);
            this.instance.callSapService();
            this.instance.afterOnblur.subscribe((/**
             * @return {?}
             */
            () => {
                e.editorRef.endPending();
            }));
        });
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        super.ngAfterViewInit();
    }
}
DatagridExtIntgrtnComponent.decorators = [
    { type: Component, args: [{
                selector: "grid-external-integration",
                template: `
    <div [formGroup]="group" class="f-datagrid-cell-formgroup farris-group-auto">
      <web-external-integration
        #extIntgrtn
        style="width: 100%"
        extIntegration-dataMapping
        [placeholder]="options.placeholder"
        [readonly]="options.readonly"
        [mapFields]="options.mapFields"
        [serviceCode] = "options.serviceCode"
        [externalParams] = "options.externalParams"
        [extTableSchemas] = "options.extTableSchemas"
        [formControlName]="column.field"
        [mappingType] = "'grid'"
        
      ></web-external-integration>
    </div>
  `
            }] }
];
/** @nocollapse */
DatagridExtIntgrtnComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: RuntimeStateService },
    { type: Injector }
];
DatagridExtIntgrtnComponent.propDecorators = {
    instance: [{ type: ViewChild, args: ["extIntgrtn",] }]
};
/** @type {?} */
const ExtIntgrtnDataGridEditorProvider = {
    provide: GRID_EDITORS,
    useValue: { name: "external-integration", value: DatagridExtIntgrtnComponent },
    multi: true
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ext-integration-data-mapping.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ExtIntegrationDataMappingDirective {
    /**
     * @param {?} vm
     * @param {?} extIntegrationComponent
     */
    constructor(vm, extIntegrationComponent) {
        this.vm = vm;
        this.extIntegrationComponent = extIntegrationComponent;
        this.showModal = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.extIntegrationComponent.selectSapData.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            this.mappingData(data, this.myField);
        }));
    }
    /**
     * 字段映射
     * @param {?} sapData sap返回信息
     * @param {?} mappingField 映射关系
     * @return {?}
     */
    mappingData(sapData, mappingField) {
        /** @type {?} */
        let helpFields = Object.keys(mappingField);
        helpFields.forEach((/**
         * @param {?} fieldName
         * @return {?}
         */
        (fieldName) => {
            /** @type {?} */
            let pathArr = this.getBindingPathArray();
            // 获取字段值
            /** @type {?} */
            let val;
            if (sapData) {
                val = sapData[fieldName];
            }
            this.vm.bindingData.setValue(pathArr.concat(mappingField[fieldName]), val, true, true);
        }));
    }
    /**
     * @private
     * @return {?}
     */
    getBindingPathArray() {
        /** @type {?} */
        const path = this.vm.bindingPath;
        if (path) {
            return path.split("/").filter((/**
             * @param {?} n
             * @return {?}
             */
            (n) => n !== ""));
        }
        return [];
    }
}
ExtIntegrationDataMappingDirective.decorators = [
    { type: Directive, args: [{
                selector: "[extIntegration-dataMapping]",
            },] }
];
/** @nocollapse */
ExtIntegrationDataMappingDirective.ctorParameters = () => [
    { type: ViewModel, decorators: [{ type: Optional }] },
    { type: ExternalIntegrationComponent }
];
ExtIntegrationDataMappingDirective.propDecorators = {
    myField: [{ type: Input, args: ["mapFields",] }],
    showModal: [{ type: Output }]
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/external-integration.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ExternalIntegrationModule {
}
ExternalIntegrationModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    ExtIntegrationDataMappingDirective,
                    ExternalIntegrationComponent,
                    DatagridExtIntgrtnComponent
                ],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    ListViewModule,
                    RouterModule,
                    ListNavModule,
                    InputGroupModule,
                    TextModule,
                    HttpClientModule,
                    FormMessageModule,
                    PerfectScrollbarModule,
                    TreeModule,
                    FarrisTabsModule,
                    FarrisFormsModule,
                    FormsModule,
                    FarrisCommonModule.forRoot(),
                    FarrisDialogModule.forRoot(),
                    MessagerModule.forRoot(),
                    NotifyModule.forRoot(),
                    LayoutModule,
                    LocaleModule.forRoot(),
                    ModalModule.forRoot(),
                    DatagridModule
                ],
                providers: [BsModalService, DialogService],
                entryComponents: [DatagridExtIntgrtnComponent, DatagridComponent],
                exports: [
                    ExtIntegrationDataMappingDirective,
                    ExternalIntegrationComponent,
                    DatagridExtIntgrtnComponent
                ]
            },] }
];

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: farris-external-integration.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { DatagridExtIntgrtnComponent, ExtIntgrtnDataGridEditorProvider, ExternalIntegrationModule, ExtIntegrationDataMappingDirective, ExternalIntegrationService, ExternalIntegrationComponent };
