// @gsp-wf/wf-process-editor/esm2015/lib/views/performer/entities/performer-info.entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class PerformerInfo {
    constructor() {
        this.performers = new Array();
        this.components = new Array();
        this.reportingLines = new Array();
        this.matrixParticipants = new Array();
        this.activityExecutors = new Array();
        this.organizationParticipants = new Array();
    }
}
if (false) {
    /** @type {?} */
    PerformerInfo.prototype.performers;
    /** @type {?} */
    PerformerInfo.prototype.variablePerformers;
    /** @type {?} */
    PerformerInfo.prototype.components;
    /** @type {?} */
    PerformerInfo.prototype.reportingLines;
    /** @type {?} */
    PerformerInfo.prototype.matrixParticipants;
    /** @type {?} */
    PerformerInfo.prototype.activityExecutors;
    /** @type {?} */
    PerformerInfo.prototype.organizationParticipants;
    /** @type {?} */
    PerformerInfo.prototype.options;
    /** @type {?} */
    PerformerInfo.prototype.isAllScope;
}
