// @gsp-svc/expression/esm5/lib/expression-var-grid/expression-var-grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { ExpressionEditorLocalePipe } from '../expression-editor-locale.pipe';
import { VarType } from './vartype';
var ExpressionVarGridComponent = /** @class */ (function () {
    function ExpressionVarGridComponent() {
        this.selectVarRow = new EventEmitter();
        this.columns = [
            { field: 'name', title: '名称' },
        ];
        this.items = [];
        //初始化i18n标识
        this.localeId = localStorage.getItem('languageCode');
        this.localePipe = new ExpressionEditorLocalePipe(this.localeId);
    }
    /**
     * @return {?}
     */
    ExpressionVarGridComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.items = [
            {
                name: this.localePipe.transform("VariableSessionValue"),
                id: VarType.Session
            },
            {
                name: this.localePipe.transform("VariableContextParameter"),
                id: VarType.Context
            }
        ];
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    ExpressionVarGridComponent.prototype.selectRowChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.selectVarRow.emit($event.data);
    };
    ExpressionVarGridComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-expression-var-grid',
                    template: "<farris-datagrid\r\n[idField]=\"'id'\"\r\n[height]=\"268\"\r\n[columns]=\"columns\" \r\n[data]=\"items\"\r\n[fit]=\"true\"\r\n[fitColumns]=\"true\"\r\n[striped]=\"false\"\r\n[showHeader]=\"false\"\r\n[headerHeight]=\"30\"\r\n[pagination]=\"false\"\r\n(selectChanged)=\"selectRowChange($event)\"\r\n>\r\n</farris-datagrid>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ExpressionVarGridComponent.ctorParameters = function () { return []; };
    ExpressionVarGridComponent.propDecorators = {
        h1: [{ type: Input }],
        selectVarRow: [{ type: Output }]
    };
    return ExpressionVarGridComponent;
}());
export { ExpressionVarGridComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ExpressionVarGridComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    ExpressionVarGridComponent.prototype.localePipe;
    /** @type {?} */
    ExpressionVarGridComponent.prototype.h1;
    /** @type {?} */
    ExpressionVarGridComponent.prototype.selectVarRow;
    /** @type {?} */
    ExpressionVarGridComponent.prototype.columns;
    /** @type {?} */
    ExpressionVarGridComponent.prototype.items;
}
