// @farris/ui-common/esm2015/column/column-format.options.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function ColumnFormatter() { }
if (false) {
    /** @type {?|undefined} */
    ColumnFormatter.prototype.type;
    /** @type {?|undefined} */
    ColumnFormatter.prototype.options;
}
/**
 * @record
 */
export function NumberFormatOptions() { }
if (false) {
    /**
     * 货币符号
     * @type {?|undefined}
     */
    NumberFormatOptions.prototype.prefix;
    /**
     * 后缀符号
     * @type {?|undefined}
     */
    NumberFormatOptions.prototype.suffix;
    /**
     * 输出格式: %s = 货币符号, %v = 值
     * @type {?|undefined}
     */
    NumberFormatOptions.prototype.format;
    /**
     * 小数点分隔符
     * @type {?|undefined}
     */
    NumberFormatOptions.prototype.decimal;
    /**
     * 千分位分隔符
     * @type {?|undefined}
     */
    NumberFormatOptions.prototype.thousand;
    /**
     * 保留小数位数
     * @type {?|undefined}
     */
    NumberFormatOptions.prototype.precision;
    /**
     * 去掉小数末尾的0
     * @type {?|undefined}
     */
    NumberFormatOptions.prototype.stripZeros;
}
/**
 * @record
 */
export function DateTimeFormatOptions() { }
if (false) {
    /** @type {?|undefined} */
    DateTimeFormatOptions.prototype.format;
}
/**
 * @record
 */
export function BooleanFormatOptions() { }
if (false) {
    /** @type {?|undefined} */
    BooleanFormatOptions.prototype.trueText;
    /** @type {?|undefined} */
    BooleanFormatOptions.prototype.falseText;
}
/**
 * @record
 */
export function EnumFormatOptions() { }
if (false) {
    /** @type {?} */
    EnumFormatOptions.prototype.valueField;
    /** @type {?} */
    EnumFormatOptions.prototype.textField;
    /** @type {?|undefined} */
    EnumFormatOptions.prototype.data;
}
/**
 * @record
 */
export function ImageFormatOptions() { }
if (false) {
    /** @type {?|undefined} */
    ImageFormatOptions.prototype.width;
    /** @type {?|undefined} */
    ImageFormatOptions.prototype.height;
}
