// @gsp-wf/wf-process-editor/esm2015/lib/views/title-expression/title-expression.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input, ViewChild, TemplateRef, Injector, LOCALE_ID, ChangeDetectorRef, Optional, Inject } from '@angular/core';
import { ExpressionService, ExpressioneditorComponent } from '@gsp-svc/expression';
import { ProcessDesignerUIState } from '../../app/process-designer.uistate';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../i18n/pipe/wf-process-editor-locale.pipe';
export class TitleExpressionComponent {
    /**
     * @param {?} expr
     * @param {?} injector
     * @param {?} designerState
     * @param {?} cdr
     * @param {?} localeId
     * @param {?} resources
     * @param {?} localePipe
     */
    constructor(expr, injector, designerState, cdr, localeId, resources, localePipe) {
        this.expr = expr;
        this.injector = injector;
        this.designerState = designerState;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.textWithMultiLanguage = new Map();
        this.editorParams = {};
        this.language = 'zh-CHS';
        this.languageOptions = [
            { value: 'zh-CHS', text: this.localePipe.transform("component.titleExpression.zh-CHS") },
            { value: 'zh-CHT', text: this.localePipe.transform("component.titleExpression.zh-CHT") },
            { value: 'en', text: this.localePipe.transform("component.titleExpression.en") }
        ];
        this.contextEntities = [];
        this.schema = {};
        this.injectName = '';
        this.modalConfig = {
            title: this.localePipe.transform("component.titleExpression.taskDescription"),
            width: 1140,
            height: 720,
            showButtons: true
        };
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
        if (this.designerState.schemas[0]) {
            this.schema = this.designerState.schemas[0];
            if (this.designerState.schemas[0].entityTypes && this.designerState.schemas[0].entityTypes.length > 0) {
                this.injectName = this.designerState.schemas[0].entityTypes[0].name;
            }
        }
        this.language = this.injector.get(LOCALE_ID, null) || this.language;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // 构造多语表达式编辑器的数据源，兼容历史
        if (this.value != null) {
            if (typeof (this.value) === 'string') {
                this.textWithMultiLanguage.set('zh-CHS', this.value);
            }
            else {
                this.textWithMultiLanguage = new Map(Object.entries(this.value));
            }
        }
        this.contextEntities = new Array();
        if (this.designerState.procInstContext.length > 0) {
            for (const variable of this.designerState.procInstContext) {
                this.contextEntities = this.contextEntities.concat([variable]);
            }
        }
        if (this.designerState.formalParameterContext.length > 0) {
            for (const variable of this.designerState.formalParameterContext) {
                if (!this.contextEntities.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                c => c.key === variable.key))) {
                    this.contextEntities = this.contextEntities.concat([variable]);
                }
            }
        }
        if (this.designerState.subActivityContext.length > 0) {
            for (const variable of this.designerState.subActivityContext) {
                if (!this.contextEntities.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                c => c.key === variable.key))) {
                    this.contextEntities = this.contextEntities.concat([variable]);
                }
            }
        }
        if (this.designerState.participantContext.length > 0) {
            for (const variable of this.designerState.participantContext) {
                if (!this.contextEntities.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                c => c.key === variable.key))) {
                    this.contextEntities = this.contextEntities.concat([variable]);
                }
            }
        }
        if (this.designerState.processVariableContext.length > 0) {
            for (const variable of this.designerState.processVariableContext) {
                if (!this.contextEntities.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                c => c.key === variable.key))) {
                    this.contextEntities = this.contextEntities.concat([variable]);
                }
            }
        }
    }
    /**
     * @return {?}
     */
    clickOK() {
        this.value = this.editor.getExprWithMultiLanguage();
        /** @type {?} */
        const obj = [...this.value.entries()].reduce((/**
         * @param {?} obj
         * @param {?} __1
         * @return {?}
         */
        (obj, [key, value]) => (obj[key] = value == null ? '' : value, obj)), {});
        /** @type {?} */
        const object = { value: obj, parameters: null };
        this.submitModal.emit(object);
    }
    /**
     * @return {?}
     */
    clickCancel() {
        this.closeModal.emit();
    }
    /**
     * @private
     * @param {?} val
     * @return {?}
     */
    convertToValue(val) {
        if (val) {
            if (val.substring(9, val.length - 2)) {
                return val;
            }
            else {
                return '';
            }
        }
        else {
            return '';
        }
    }
}
TitleExpressionComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-title-expression',
                template: "<div class=\"f-page\">\r\n  <div class=\"f-page-header\">\r\n    <nav class=\"f-page-header-base\">\r\n      <div class=\"f-title\">\r\n        <div class=\"col-12 farris-group-auto pr-5\">\r\n          <div class=\"farris-group-wrap\">\r\n            <div class=\"form-group farris-form-group\" style=\"display:flex\">\r\n              <label class=\"col-form-label\">\r\n                <span class=\"farris-label-text\">{{'component.titleExpression.currentLanguage' | wfLocale}}</span>\r\n              </label>\r\n              <div class=\"farris-input-wrap\">\r\n                <farris-combo-list [data]=\"languageOptions\" [textField]=\"'text'\" [valueField]=\"'value'\"\r\n                  [idField]=\"'value'\" [(ngModel)]=\"language\" style=\"flex: 1;\" [enableClear]=\"false\">\r\n                </farris-combo-list>\r\n                <div class=\"farris-feedback valid-feedback\"></div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </nav>\r\n  </div>\r\n  <div class=\"f-page-main f-utils-fill-flex-column h-100\">\r\n    <div class=\"f-utils-fill-flex-column farris-overflow-y-auto h-100\">\r\n      <app-expressioneditor class=\"h-100\" [isPanel]=\"true\" [contextEntities]=\"contextEntities\" [text]=\"value\"\r\n        [customSchemaName]=\"'Schema'\" [customSchema]=\"schema\" [injectName]=\"injectName\" [injectSchemaName]=\"'Schema'\"\r\n        [injectNameInSchema]=\"injectName\" [language]=\"language\" [textWithMultiLanguage]=\"textWithMultiLanguage\">\r\n      </app-expressioneditor>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #iconSelectFooter>\r\n  <div class=\"btns\">\r\n    <button type=\"button\" class=\"btn btn-secondary btn-sm\" (click)=\"clickCancel()\">{{'component.titleExpression.clickCancel' | wfLocale}}</button>\r\n    <button type=\"button\" class=\"btn btn-primary btn-sm mr-2\" (click)=\"clickOK()\">{{'component.titleExpression.clickOK' | wfLocale}}</button>\r\n  </div>\r\n</ng-template>\r\n",
                providers: [
                    WfLocalePipe,
                ],
                styles: [""]
            }] }
];
/** @nocollapse */
TitleExpressionComponent.ctorParameters = () => [
    { type: ExpressionService },
    { type: Injector },
    { type: ProcessDesignerUIState },
    { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
    { type: WfLocalePipe, decorators: [{ type: Optional }] }
];
TitleExpressionComponent.propDecorators = {
    closeModal: [{ type: Output }],
    submitModal: [{ type: Output }],
    value: [{ type: Input }],
    modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }],
    editorParams: [{ type: Input }],
    editor: [{ type: ViewChild, args: [ExpressioneditorComponent,] }]
};
if (false) {
    /** @type {?} */
    TitleExpressionComponent.prototype.closeModal;
    /** @type {?} */
    TitleExpressionComponent.prototype.submitModal;
    /** @type {?} */
    TitleExpressionComponent.prototype.value;
    /** @type {?} */
    TitleExpressionComponent.prototype.textWithMultiLanguage;
    /** @type {?} */
    TitleExpressionComponent.prototype.modalFooter;
    /** @type {?} */
    TitleExpressionComponent.prototype.editorParams;
    /** @type {?} */
    TitleExpressionComponent.prototype.language;
    /** @type {?} */
    TitleExpressionComponent.prototype.editor;
    /** @type {?} */
    TitleExpressionComponent.prototype.languageOptions;
    /** @type {?} */
    TitleExpressionComponent.prototype.contextEntities;
    /** @type {?} */
    TitleExpressionComponent.prototype.schema;
    /** @type {?} */
    TitleExpressionComponent.prototype.injectName;
    /** @type {?} */
    TitleExpressionComponent.prototype.modalConfig;
    /**
     * @type {?}
     * @private
     */
    TitleExpressionComponent.prototype.expr;
    /**
     * @type {?}
     * @private
     */
    TitleExpressionComponent.prototype.injector;
    /** @type {?} */
    TitleExpressionComponent.prototype.designerState;
    /**
     * @type {?}
     * @private
     */
    TitleExpressionComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    TitleExpressionComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    TitleExpressionComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    TitleExpressionComponent.prototype.localePipe;
}
export class TitleExpressionConverter {
    /**
     * @param {?} localePipe
     */
    constructor(localePipe) {
        this.localePipe = localePipe;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    convertTo(data) {
        if (data && typeof data === 'string') {
            return data;
        }
        else if (data && Object.keys(data).length > 0) {
            return this.areAllPropertiesEmpty(data)
                ? this.localePipe.transform("component.titleExpression.notSet")
                : this.localePipe.transform("component.titleExpression.set");
        }
        else {
            return this.localePipe.transform("component.titleExpression.notSet");
        }
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    areAllPropertiesEmpty(obj) {
        for (let key in obj) {
            if (obj.hasOwnProperty(key) && obj[key] !== '') {
                return false;
            }
        }
        return true;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    TitleExpressionConverter.prototype.localePipe;
}
