// @gsp-wf/wf-process-model/esm2015/lib/model/mex/CDTSchemaType.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, CustomType } from "@edp-pmf/bpmn-model";
export class CDTSchemaType extends CustomType {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = CDTSchemaType.Clr_Type_ID;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new CDTSchemaType(this.Model);
    }
    /**
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.MetadataId = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "SchemaId", this.SchemaId);
        BpmnModelHelper.SetAttributeValue(obj, "MetadataId", this.MetadataId);
        BpmnModelHelper.SetAttributeValue(obj, "ComponentId", this.ComponentId);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.SchemaId = obj.SchemaId;
        this.MetadataId = obj.MetadataId;
        this.ComponentId = obj.ComponentId;
    }
}
CDTSchemaType.Clr_Type_ID = "wf.CDTSchemaType";
if (false) {
    /** @type {?} */
    CDTSchemaType.Clr_Type_ID;
    /** @type {?} */
    CDTSchemaType.prototype.ClrTypeID;
    /** @type {?} */
    CDTSchemaType.prototype.SchemaId;
    /** @type {?} */
    CDTSchemaType.prototype.MetadataId;
    /** @type {?} */
    CDTSchemaType.prototype.ComponentId;
}
BpmnModelHelper.RegisterElementType(new CDTSchemaType(new BpmnModel()));
