// @gsp-wf/wf-process-model/esm2015/lib/model/mex/MexExpressionImpl.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, MexExpression } from "@edp-pmf/bpmn-model";
export class MexExpressionImpl extends MexExpression {
    /**
     * @param {?} model
     * @param {?=} value
     */
    constructor(model, value) {
        super(model);
        this.ClrTypeID = MexExpressionImpl.Clr_Type_ID;
        this.value = value;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.value = '';
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, 'Value', this.value);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.value = obj.Value;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new MexExpressionImpl(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const mexExpressionImpl = new MexExpressionImpl(this.Model);
        mexExpressionImpl.value = this.value;
        return mexExpressionImpl;
    }
}
MexExpressionImpl.Clr_Type_ID = "wf.MexExpressionImpl";
if (false) {
    /** @type {?} */
    MexExpressionImpl.Clr_Type_ID;
    /** @type {?} */
    MexExpressionImpl.prototype.ClrTypeID;
    /** @type {?} */
    MexExpressionImpl.prototype.value;
}
BpmnModelHelper.RegisterElementType(new MexExpressionImpl(new BpmnModel()));
