// @gsp-wf/wf-process-model/esm2015/lib/model/mex/TitleExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { MexExpressionImpl } from "./MexExpressionImpl";
export class TitleExpression extends MexExpressionImpl {
    /**
     * @param {?} model
     * @param {?=} value
     * @param {?=} valueLanguage
     */
    constructor(model, value, valueLanguage) {
        super(model, value);
        this.ClrTypeID = TitleExpression.Clr_Type_ID;
        this.valueLanguage = {};
        this.valueLanguage = valueLanguage;
    }
    /**
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.valueLanguage = {};
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, 'ValueLanguage', this.valueLanguage);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.valueLanguage = obj.ValueLanguage;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new TitleExpression(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const titleExpression = new TitleExpression(this.Model);
        titleExpression.value = this.value;
        titleExpression.valueLanguage = this.valueLanguage;
        return titleExpression;
    }
}
TitleExpression.Clr_Type_ID = 'wf.TitleExpression';
if (false) {
    /** @type {?} */
    TitleExpression.Clr_Type_ID;
    /** @type {?} */
    TitleExpression.prototype.ClrTypeID;
    /** @type {?} */
    TitleExpression.prototype.valueLanguage;
}
BpmnModelHelper.RegisterElementType(new TitleExpression(new BpmnModel()));
