// @gsp-wf/wf-dimension-def/esm5/lib/refModule/panel/panel-item/panel-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ContentChild, TemplateRef, Optional, ViewEncapsulation, ChangeDetectorRef, HostBinding, Renderer2, ElementRef } from '@angular/core';
import { dropAnimation } from '../panel/util/drop.animation';
import { WFDimensionPanelComponent } from '../panel/panel-item-list.component';
var WFDimensionPanelItemComponent = /** @class */ (function () {
    function WFDimensionPanelItemComponent(panel, changeRef, render, el) {
        this.panel = panel;
        this.changeRef = changeRef;
        this.render = render;
        this.el = el;
        this.sysInit = false;
        // @HostBinding('class.farris-panel-item') isPanelItem = true;
        // @HostBinding('class.card') isCard = true;
        // 默认非激活状态(不显示)
        this.active = false;
    }
    /**
     * @return {?}
     */
    WFDimensionPanelItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    WFDimensionPanelItemComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.editState) {
            if (this.editState) {
                this.active = true;
            }
            else {
                this.active = false;
            }
        }
    };
    WFDimensionPanelItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-dimension-panel-item',
                    template: "<!-- \u6807\u9898  \u662F\u5426\u6709\u6A21\u677F -->\r\n\r\n<div id=\"cardheader\" class=\"card-header\" style=\"height: 45px; background-color: white;border-bottom: 1px solid rgba(0, 0, 0, .125);\">\r\n    <!-- \u6807\u9898  \u662F\u5426\u6709\u6A21\u677F -->\r\n    <div class=\"panel-item-title\">\r\n        <span class=\"f-icon f-icon-new-dimension\"></span>\r\n        <span *ngIf=\"!headRef;else head\">\r\n            {{tabTitle}}\r\n        </span>\r\n        <span [class]=\"sysInit?'sysInit-flow-form':'customize-flow-form'\">{{sysInit?'\u7CFB\u7EDF\u9884\u7F6E':'\u81EA\u5B9A\u4E49'}}</span>\r\n        <ng-template #head [ngTemplateOutlet]=\"headRef\"></ng-template>\r\n    </div>\r\n    <div class=\"panel-item-tool\">\r\n        <ng-template [ngTemplateOutlet]=\"toolTempl\"></ng-template>\r\n    </div>\r\n    <div class=\"panel-item-clear\">\r\n    </div>\r\n\r\n</div>\r\n<!-- panel\u5185\u5BB9\u6A21\u677F -->\r\n<div class=\" card-body \">\r\n    <ng-template [ngTemplateOutlet]=\"contentRef\">\r\n    </ng-template>\r\n</div>\r\n<!-- \u5E95\u90E8  \u662F\u5426\u6709\u6A21\u677F -->\r\n<div class=\"card-bottom \" style=\"border-top: 1px solid rgba(0, 0, 0, .125);\" [@dropAnimation]=\"active? 'active': 'inactive' \">\r\n    <div style=\"float: right; margin: 9px;\">\r\n        <ng-template [ngTemplateOutlet]=\"bottomRef \">\r\n        </ng-template>\r\n    </div>\r\n</div>",
                    animations: [dropAnimation],
                    encapsulation: ViewEncapsulation.None,
                    styles: [":host{width:100%}#cardheader .card-header{cursor:pointer;border-width:1px;background-color:#fff}.panel-item-title{float:left}.card-bottom{background-color:rgba(242,242,242);height:45px}.panel-item-tool{float:right}.panel-item-clear{clear:both}.f-state-disable{pointer-events:none}.customize-flow-form{font-size:12px;color:#fff;margin-left:10px;border-radius:20px;padding:2px 8px;background:#59a1ff}.sysInit-flow-form{font-size:12px;color:#fff;margin-left:10px;border-radius:20px;padding:2px 8px;background:#ff8e59}"]
                }] }
    ];
    /** @nocollapse */
    WFDimensionPanelItemComponent.ctorParameters = function () { return [
        { type: WFDimensionPanelComponent, decorators: [{ type: Optional }] },
        { type: ChangeDetectorRef },
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    WFDimensionPanelItemComponent.propDecorators = {
        editState: [{ type: Input }],
        width: [{ type: Input }],
        height: [{ type: Input }],
        value: [{ type: Input }],
        tabTitle: [{ type: Input, args: ['title',] }],
        sysInit: [{ type: Input }],
        disable: [{ type: Input }, { type: HostBinding, args: ['class.f-state-disable',] }],
        headRef: [{ type: ContentChild, args: ['headTempl',] }],
        bottomRef: [{ type: ContentChild, args: ['bottomRef',] }],
        toolTempl: [{ type: ContentChild, args: ['toolTempl',] }],
        contentRef: [{ type: ContentChild, args: ['contentTempl',] }]
    };
    return WFDimensionPanelItemComponent;
}());
export { WFDimensionPanelItemComponent };
if (false) {
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.editState;
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.width;
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.height;
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.value;
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.tabTitle;
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.sysInit;
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.disable;
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.headRef;
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.bottomRef;
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.toolTempl;
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.contentRef;
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.active;
    /** @type {?} */
    WFDimensionPanelItemComponent.prototype.panel;
    /**
     * @type {?}
     * @private
     */
    WFDimensionPanelItemComponent.prototype.changeRef;
    /**
     * @type {?}
     * @private
     */
    WFDimensionPanelItemComponent.prototype.render;
    /**
     * @type {?}
     * @private
     */
    WFDimensionPanelItemComponent.prototype.el;
}
