// @gsp-wf/wf-dimension-def/esm5/gsp-wf-wf-dimension-def.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { WfDimensionDefService, WfDimensionDefComponent, WfDimensionDefModule } from './public-api';
export { DimensionItemCardComponent as ɵc } from './lib/dimension-item-card/dimension-item-card.component';
export { EnumEditComponent as ɵd } from './lib/enum-edit/enum-edit.component';
export { WFDimensionPanelItemComponent as ɵg } from './lib/refModule/panel/panel-item/panel-item.component';
export { WFDimensionPanelModule as ɵe } from './lib/refModule/panel/panel.module';
export { WFDimensionPanelComponent as ɵf } from './lib/refModule/panel/panel/panel-item-list.component';
export { dropAnimation as ɵh } from './lib/refModule/panel/panel/util/drop.animation';
export { SelectElementComponent as ɵj } from './lib/refModule/select-element/select-element.component';
export { SelectElementModule as ɵi } from './lib/refModule/select-element/select-element.module';
export { SelectObjElementComponent as ɵl } from './lib/refModule/select-element/steps/select-obj-element/select-obj-element.component';
export { SelectObjectComponent as ɵk } from './lib/refModule/select-element/steps/select-object/select-object.component';
export { WFDimensionDefUIState as ɵb } from './lib/service/wf-dimension-def.uistate';
export { WFDimensionDefViewmodel as ɵa } from './lib/service/wf-dimension-def.viewModel';
