// @farris/extend-file-upload/esm5/lib/service/f-file-upload-api.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/f-file-upload-api.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional, Inject, InjectionToken } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function FFileUploadServerConfig() { }
if (false) {
    /** @type {?|undefined} */
    FFileUploadServerConfig.prototype.upload;
    /** @type {?|undefined} */
    FFileUploadServerConfig.prototype.remove;
}
/** @type {?} */
export var FFileUploadAPIServiceToken = new InjectionToken('MFFileUploadServerConfig');
var FFileUploadAPIService = /** @class */ (function () {
    function FFileUploadAPIService(exConfig) {
        var _this = this;
        this.apiArray = ['upload', 'remove'];
        this.config = { upload: null, remove: null };
        this.commonAPI = { type: 'config', url: '', timeout: 0, headers: null, data: {} };
        if (exConfig) {
            for (var key in this.commonAPI) {
                if (exConfig.hasOwnProperty(key)) {
                    this.commonAPI[key] = exConfig[key];
                }
            }
        }
        this.apiArray.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            _this.config[item] = Object.assign({}, _this.commonAPI);
            if (exConfig && exConfig.hasOwnProperty(item)) {
                if (exConfig[item].hasOwnProperty('url')) {
                    _this.config[item] = Object.assign({}, _this.config[item], exConfig[item], { url: _this.combinUrl(exConfig[item].url) });
                }
            }
        }));
        // if (exConfig) {
        //     Object.assign(this.config, exConfig);
        // }
    }
    /**
     * @private
     * @param {?} typeUrl
     * @return {?}
     */
    FFileUploadAPIService.prototype.combinUrl = /**
     * @private
     * @param {?} typeUrl
     * @return {?}
     */
    function (typeUrl) {
        /** @type {?} */
        var result = typeUrl;
        result = this.commonAPI.url + typeUrl;
        return result;
    };
    /**
     * 获取API定义
     */
    /**
     * 获取API定义
     * @return {?}
     */
    FFileUploadAPIService.prototype.getAPI = /**
     * 获取API定义
     * @return {?}
     */
    function () {
        return this.config;
    };
    /**
     *
     * @param type
     * @param object
     */
    /**
     *
     * @param {?} type
     * @param {?} object
     * @return {?}
     */
    FFileUploadAPIService.prototype.getAssignAPI = /**
     *
     * @param {?} type
     * @param {?} object
     * @return {?}
     */
    function (type, object) {
        // removeAll,uploadAll 用于内部传递使用，应当传递的是remove和upload对应的参数
        // 通过设置对应关系
        /** @type {?} */
        var infoMap = { 'removeAll': 'remove', 'uploadAll': 'upload', 'multipartUpload': 'upload' };
        if (infoMap.hasOwnProperty(type)) {
            type = infoMap[type];
        }
        if (this.config[type] && object) {
            return Object.assign({}, this.config[type], object);
        }
        if (this.config[type]) {
            return this.config[type];
        }
        if (object) {
            return object;
        }
        return null;
    };
    FFileUploadAPIService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    FFileUploadAPIService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [FFileUploadAPIServiceToken,] }] }
    ]; };
    /** @nocollapse */ FFileUploadAPIService.ngInjectableDef = i0.defineInjectable({ factory: function FFileUploadAPIService_Factory() { return new FFileUploadAPIService(i0.inject(FFileUploadAPIServiceToken, 8)); }, token: FFileUploadAPIService, providedIn: "root" });
    return FFileUploadAPIService;
}());
export { FFileUploadAPIService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FFileUploadAPIService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    FFileUploadAPIService.prototype.commonAPI;
    /**
     * @type {?}
     * @private
     */
    FFileUploadAPIService.prototype.apiArray;
}
