// @farris/command-services/esm2015/lib/data-services/edit-data.service.js
import { Injectable } from '@angular/core';
import { FrameContext } from '@farris/devkit';
import { EMPTY } from 'rxjs';
import { tap } from 'rxjs/operators';
import { FormLoadingService } from '../form-loading/form-loading.service';
import { LanguageService } from '../languag.service';
import { FormErrorService } from '../error/form-error.service';
import { EditStateUtil } from './utils/index';
/**
 * 数据编辑服务
 */
class EditDataService {
    /**
     * 构造函数
     */
    constructor(frameContext) {
        this.frameContext = frameContext;
        this.repository = this.frameContext.repository;
        this.bindingData = this.frameContext.bindingData;
        this.loadingService = this.frameContext.injector.get(FormLoadingService, null);
        this.languageService = this.frameContext.injector.get(LanguageService, null);
        this.errorService = this.frameContext.injector.get(FormErrorService, null);
    }
    /**
     * 编辑（支持加锁）
     */
    edit(id) {
        id = id || this.bindingData.list.currentId;
        if (!id) {
            return EMPTY;
        }
        const loadingTimerId = this.loadingService.showLoadingWithDelay(500);
        const edit$ = this.repository.editEntityById(id);
        const result$ = edit$.pipe(tap(() => {
            this.loadingService.hideDelayLoading(loadingTimerId);
            EditStateUtil.setEditState(this.frameContext, id);
        }, (error) => {
            this.loadingService.hideDelayLoading(loadingTimerId);
            this.errorService.exception(this.languageService.updateFailed, error);
        }));
        return result$;
    }
}
EditDataService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
EditDataService.ctorParameters = () => [
    { type: FrameContext }
];
export { EditDataService };
