// @gsp-wf/wf-dimension-def/esm2015/lib/dimension-item-card/dimension-item-card.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ComponentFactoryResolver, Injector, EventEmitter, Input } from '@angular/core';
import { BsModalService } from '@farris/ui-modal';
import { MessagerService } from '@farris/ui-messager';
import { DimensionDefEntity, WFDimensionTypes, WFDimensionInputTypes, DimensionInputType } from '../entity/dimensiondef.entity';
import { WFDimensionDefUIState } from '../service/wf-dimension-def.uistate';
import { ExpressionService, InjectedType } from '@gsp-svc/expression';
import { EnumEditComponent } from '../enum-edit/enum-edit.component';
import { LookupConfigService } from '@farris/lookup-config';
export class DimensionItemCardComponent {
    /**
     * @param {?} modalService
     * @param {?} resolver
     * @param {?} injector
     * @param {?} messageService
     * @param {?} uistate
     * @param {?} exprService
     * @param {?} configService
     */
    constructor(modalService, resolver, injector, messageService, uistate, exprService, configService) {
        this.modalService = modalService;
        this.resolver = resolver;
        this.injector = injector;
        this.messageService = messageService;
        this.uistate = uistate;
        this.exprService = exprService;
        this.configService = configService;
        this.sysInit = false;
        this.canEdit = false;
        /**
         * 维度种类
         */
        this.dimensionTypes = WFDimensionTypes;
        this.inputTypes = WFDimensionInputTypes;
        this.selectedBEMetadata = new EventEmitter();
        this.selectedHelpMetadata = new EventEmitter();
        this.groupIcon = '<i class="f-icon f-icon-lookup"></i>';
        this.enumListText = (/**
         * @return {?}
         */
        () => {
            if (!this.getIsEnumInput()) {
                return '';
            }
            /** @type {?} */
            let display = '';
            if (this.data.enumList) {
                /** @type {?} */
                const enumList = JSON.parse(this.data.enumList);
                if (enumList.length > 0) {
                    enumList.forEach((/**
                     * @param {?} item
                     * @return {?}
                     */
                    item => {
                        display = display.concat(`${item.key}:${item.value}; `);
                    }));
                }
            }
            return display;
        });
        this.configService.getLookupData().subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            console.log(data);
            this.data.smartHelp = data;
        }));
    }
    /**
     * @return {?}
     */
    get dimensionItemCode() {
        return this.data ? this.data.code : '';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set dimensionItemCode(value) {
        this.data.code = value;
    }
    /**
     * @return {?}
     */
    get dimensionItemName() {
        return this.data ? this.data.dimensionName : '';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set dimensionItemName(value) {
        this.data.dimensionName = value;
    }
    /**
     * @return {?}
     */
    get dimensionCode() {
        return this.data ? this.data.dimensionCode : '';
    }
    /**
     * @param {?} v
     * @return {?}
     */
    set dimensionCode(v) {
        this.data.dimensionCode = v;
    }
    /**
     * @return {?}
     */
    get inputType() {
        return this.data ? this.data.inputType : null;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    set inputType(v) {
        this.data.inputType = v;
    }
    /**
     * @return {?}
     */
    get dimensionType() {
        return this.data ? this.data.dimensionType : null;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    set dimensionType(v) {
        this.data.dimensionType = v;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.canEdit = this.canCardEdit;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.canCardEdit) {
            this.canEdit = this.canCardEdit;
        }
    }
    /**
     * 维度类型
     * @param {?} data
     * @return {?}
     */
    selectDimensionType(data) {
    }
    /**
     * 维度选择1-be字段
     * @return {?}
     */
    selectBEMeta() {
        if (this.uistate.schema) {
            this.exprService.clearInjectedEntity();
            this.exprService.addSchema('Schema', this.uistate.schema);
            if (this.uistate.schema.entityTypes && this.uistate.schema.entityTypes.length > 0) {
                /** @type {?} */
                const entityType = this.uistate.schema.entityTypes[0];
                this.exprService.addInjectedEntity(entityType.name, 'Schema', entityType.name, InjectedType.EntityType);
            }
        }
        // if (this.dimensionCode) {                                 // 编辑器加载初始值
        //   try {
        this.exprService.addExpressionText(this.dimensionCode);
        // } catch (error) {
        //   console.log('parse expression error:' + this.dimensionCode);
        // }
        // }
        this.exprService.buildExpression().then((/**
         * @param {?} d
         * @return {?}
         */
        (d) => {
            if (d) {
                this.dimensionCode = d;
            }
        }));
    }
    /**
     * 输入类型
     * @param {?} data
     * @return {?}
     */
    selectInputType(data) {
        this.data.inputType = data.data.value;
    }
    /**
     * 是否枚举输入
     * @return {?}
     */
    getIsEnumInput() {
        return this.data && this.data.inputType === DimensionInputType.Enum;
    }
    /**
     * 是否帮助输入
     * @return {?}
     */
    getIsSmartHelpInput() {
        return this.data && this.data.inputType === DimensionInputType.Help;
    }
    /**
     * 枚举显示值
     * @return {?}
     */
    getEnumInputDisplay() {
        // if (!this.getIsEnumInput()) {
        //   return '';
        // }
        // const enumInputInfo = this.data.inputInfo as EnumInputInfo;
        // let display = '';
        // if (!enumInputInfo || !enumInputInfo.enumItems) {
        //   return display;
        // }
        // enumInputInfo.enumItems.forEach(item => {
        //   display = display.concat(`${item.key}:${item.value}; `);
        // });
        // return display;
    }
    /**
     * 帮助名称
     * @return {?}
     */
    get helperName() {
        if (!this.getIsSmartHelpInput()) {
            return '';
        }
        return this.data.smartHelp;
    }
    /**
     * 帮助选择
     * @return {?}
     */
    selectHelpMeta() {
        this.lookupConfig = this.configService.showConfig({
            configId: this.data.smartHelp
        });
        // const compFactory =
        //   this.resolver.resolveComponentFactory(SelectMetadataRTComponent);
        // const compRef = compFactory.create(this.injector);
        // compRef.instance.typeList = 'HelpMetadata';
        // compRef.instance.selectedMetadata
        //   .subscribe(($event: Metadata4RefDto) => {
        //     const meta: MetadataDto = $event.metadata;
        //     const help = JSON.parse(meta.content);
        //     this.data.smartHelp = help.id;
        //   });
        // const modalConfig = {
        //   title: '帮助选择',
        //   width: 930,
        //   height: 573,
        //   showButtons: false
        // };
        // const dialog = this.modalService.show(compRef, modalConfig);
        // compRef.instance.closeModalEmitter.subscribe($event => dialog.close());
    }
    /**
     * @return {?}
     */
    editEnumValues() {
        /** @type {?} */
        const compFactory = this.resolver.resolveComponentFactory(EnumEditComponent);
        /** @type {?} */
        const compRef = compFactory.create(this.injector);
        /** @type {?} */
        const modalConfig = {
            title: '枚举设置',
            width: 650,
            height: 450,
            showButtons: true,
            buttons: compRef.instance.btnRef,
        };
        if (this.data.enumList) {
            compRef.instance.load(JSON.parse(this.data.enumList));
        }
        compRef.instance.confirm.subscribe((/**
         * @param {?} re
         * @return {?}
         */
        re => {
            this.data.enumList = JSON.stringify(re) ? JSON.stringify(re) : null;
        }));
        /** @type {?} */
        const dialog = this.modalService.show(compRef, modalConfig);
        compRef.instance.closeModalEmitter.subscribe((/**
         * @return {?}
         */
        () => {
            dialog.close();
        }));
    }
}
DimensionItemCardComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf-dimension-item-card',
                template: "<div class=\"farris-form farris-form-controls-inline mt-3\">\r\n    <form class=\"f-section-formgroup\">\r\n        <div class=\"f-section-formgroup-inputs\">\r\n            <div class=\"col-12 row\">\r\n                <div class=\"farris-group-wrap col-4\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                        <label class=\"col-form-label\">\r\n                            <span class=\"farris-label-info text-danger\">*</span>\r\n                            <span class=\"farris-label-text\">\u7EF4\u5EA6\u7F16\u53F7</span>\r\n                        </label>\r\n                        <div class=\"farris-input-wrap\">\r\n                            <input-group [(ngModel)]=\"dimensionItemCode\" name=\"code\" \r\n                            [disabled]=\"!canEdit\" [readonly]=\"sysInit\" [enableClear]=\"false\"></input-group>\r\n                            <!-- <input type=\"text\" class=\"form-control\" [(ngModel)]=\"dimensionItemCode\" name=\"code\"\r\n                                [disabled]=\"!canEdit\" [readOnly]=\"canEdit && sysInit\"> -->\r\n                            <div class=\"farris-feedback valid-feedback\"></div>\r\n                        </div>\r\n                    </div>\r\n                </div>\r\n                <div class=\"farris-group-wrap col-4\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                        <label class=\"col-form-label\">\r\n                            <span class=\"farris-label-info text-danger\">*</span>\r\n                            <span class=\"farris-label-text\">\u7EF4\u5EA6\u7C7B\u578B</span>\r\n                        </label>\r\n                        <div class=\"farris-input-wrap\">\r\n                            <farris-combo-list [ngModel]=\"dimensionType\" [idField]=\"'value'\" [textField]=\"'text'\" \r\n                                [disabled]=\"!canEdit\" [enableClear]=\"false\" [readonly]=\"sysInit\"\r\n                                [data]=\"dimensionTypes\" name=\"dimensionType\" (selectChange)=\"selectDimensionType($event)\"></farris-combo-list>\r\n                            <div class=\"farris-feedback valid-feedback\"></div>\r\n                        </div>\r\n                    </div>\r\n                </div>\r\n                <div class=\"farris-group-wrap col-4\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                        <label class=\"col-form-label\">\r\n                            <span class=\"farris-label-info text-danger\">*</span>\r\n                            <span class=\"farris-label-text\">\u5206\u914D\u7EF4\u5EA6</span>\r\n                        </label>\r\n                        <div class=\"farris-input-wrap\">\r\n                            <input-group class=\"form-control\" style=\"width: 100%;padding:0;border:0;\" [enableClear]=\"false\" \r\n                                [showButtonWhenDisabled]=\"false\" [disabled]=\"!canEdit\" [readonly]=\"sysInit\"\r\n                                [ngModel]=\"dimensionCode\" name=\"dimensionCode\"\r\n                                [editable]=\"false\" (clickHandle)=\"selectBEMeta()\" [groupText]=\"groupIcon\"></input-group>\r\n                            <div class=\"farris-feedback valid-feedback\"></div>\r\n                        </div>\r\n                    </div>\r\n                </div>\r\n                <div class=\"farris-group-wrap col-4\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                        <label class=\"col-form-label\">\r\n                            <span class=\"farris-label-info text-danger\">*</span>\r\n                            <span class=\"farris-label-text\">\u7EF4\u5EA6\u540D\u79F0</span>\r\n                        </label>\r\n                        <div class=\"farris-input-wrap\">\r\n                            <input-group [(ngModel)]=\"dimensionItemName\" name=\"name\" \r\n                            [disabled]=\"!canEdit\" [readonly]=\"sysInit\" [enableClear]=\"false\"></input-group>\r\n                            <!-- <input type=\"text\" class=\"form-control\" [(ngModel)]=\"dimensionItemName\" name=\"name\" \r\n                            [disabled]=\"!canEdit\" [readOnly]=\"canEdit && sysInit\"> -->\r\n                            <div class=\"farris-feedback valid-feedback\"></div>\r\n                        </div>\r\n                    </div>\r\n                </div>\r\n                <div class=\"farris-group-wrap col-4\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                        <label class=\"col-form-label\">\r\n                            <span class=\"farris-label-info text-danger\">*</span>\r\n                            <span class=\"farris-label-text\">\u6570\u636E\u7C7B\u578B</span>\r\n                        </label>\r\n                        <div class=\"farris-input-wrap\">\r\n                            <farris-combo-list [ngModel]=\"inputType\" [idField]=\"'value'\" [textField]=\"'text'\" [disabled]=\"!canEdit\" \r\n                                [data]=\"inputTypes\" [enableClear]=\"false\" [readonly]=\"sysInit\"\r\n                                (selectChange)=\"selectInputType($event)\" name=\"inputType\"></farris-combo-list>\r\n                        </div>\r\n                    </div>\r\n                </div>\r\n                <div *ngIf=\"getIsEnumInput()\" class=\"farris-group-wrap col-4\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                        <label class=\"col-form-label\">\r\n                            <span class=\"farris-label-info text-danger\">*</span>\r\n                            <span class=\"farris-label-text\">\u679A\u4E3E\u5217\u8868</span>\r\n                        </label>\r\n                        <div class=\"farris-input-wrap\">\r\n                            <input-group class=\"form-control\" style=\"width: 100%;padding:0;border:0;\" [enableClear]=\"false\"\r\n                                [showButtonWhenDisabled]=\"false\" [disabled]=\"!canEdit\" [value]=\"enumListText()\" [editable]=\"false\"\r\n                                (clickHandle)=\"editEnumValues()\" [groupText]=\"groupIcon\" [readonly]=\"sysInit\"></input-group>\r\n                            <div class=\"farris-feedback valid-feedback\"></div>\r\n                        </div>\r\n                    </div>\r\n                </div>\r\n                <div *ngIf=\"getIsSmartHelpInput()\" class=\"farris-group-wrap col-4\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                        <label class=\"col-form-label\">\r\n                            <span class=\"farris-label-info text-danger\">*</span>\r\n                            <span class=\"farris-label-text\">\u5E2E\u52A9\u9009\u62E9</span>\r\n                        </label>\r\n                        <div class=\"farris-input-wrap\">\r\n                            <input-group class=\"form-control\" style=\"width: 100%;padding:0;border:0;\" [enableClear]=\"false\" \r\n                                [showButtonWhenDisabled]=\"false\" [disabled]=\"!canEdit\"\r\n                                [value]=\"helperName\" [editable]=\"false\" (clickHandle)=\"selectHelpMeta()\"\r\n                                [groupText]=\"groupIcon\"></input-group>\r\n                            <div class=\"farris-feedback valid-feedback\"></div>\r\n                        </div>\r\n                    </div>\r\n                </div>\r\n            </div>\r\n        </div>\r\n    </form>\r\n</div>",
                providers: [LookupConfigService],
                styles: [""]
            }] }
];
/** @nocollapse */
DimensionItemCardComponent.ctorParameters = () => [
    { type: BsModalService },
    { type: ComponentFactoryResolver },
    { type: Injector },
    { type: MessagerService },
    { type: WFDimensionDefUIState },
    { type: ExpressionService },
    { type: LookupConfigService }
];
DimensionItemCardComponent.propDecorators = {
    data: [{ type: Input }],
    canCardEdit: [{ type: Input }],
    sysInit: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DimensionItemCardComponent.prototype.data;
    /** @type {?} */
    DimensionItemCardComponent.prototype.canCardEdit;
    /** @type {?} */
    DimensionItemCardComponent.prototype.sysInit;
    /** @type {?} */
    DimensionItemCardComponent.prototype.canEdit;
    /**
     * 维度种类
     * @type {?}
     */
    DimensionItemCardComponent.prototype.dimensionTypes;
    /** @type {?} */
    DimensionItemCardComponent.prototype.inputTypes;
    /** @type {?} */
    DimensionItemCardComponent.prototype.selectedBEMetadata;
    /** @type {?} */
    DimensionItemCardComponent.prototype.selectedHelpMetadata;
    /** @type {?} */
    DimensionItemCardComponent.prototype.groupIcon;
    /** @type {?} */
    DimensionItemCardComponent.prototype.lookupConfig;
    /** @type {?} */
    DimensionItemCardComponent.prototype.enumListText;
    /**
     * @type {?}
     * @private
     */
    DimensionItemCardComponent.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    DimensionItemCardComponent.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    DimensionItemCardComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    DimensionItemCardComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    DimensionItemCardComponent.prototype.uistate;
    /**
     * @type {?}
     * @private
     */
    DimensionItemCardComponent.prototype.exprService;
    /**
     * @type {?}
     * @private
     */
    DimensionItemCardComponent.prototype.configService;
}
