// @gsp-svc/cloudprint/esm5/lib/cloudprint-formatlist/formatlist.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, LOCALE_ID, Inject } from '@angular/core';
import { NotifyService } from '@farris/ui-notify';
var FormatlistComponent = /** @class */ (function () {
    function FormatlistComponent(notifyService, localeid) {
        this.notifyService = notifyService;
        this.print = new EventEmitter();
        // dlg:BsModalRef;
        this.columns = [];
        debugger;
        // localeid = localStorage.getItem("languageCode") || localeid;
        // this.localepip = new LocalLangPipe(localeid);
    }
    /**
     * @return {?}
     */
    FormatlistComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.columns = [
            { field: 'code', title: this.localepip.transform('formatcode') },
            { field: 'name', title: this.localepip.transform('formatname') },
        ];
    };
    /**
     * @param {?} rowdata
     * @return {?}
     */
    FormatlistComponent.prototype.selectChange = /**
     * @param {?} rowdata
     * @return {?}
     */
    function (rowdata) {
        this.formatId = rowdata.id;
    };
    /**
     * @return {?}
     */
    FormatlistComponent.prototype.printJob = /**
     * @return {?}
     */
    function () {
        if (this.checkBeforePrint()) {
            this.print.emit({
                id: this.formatId
            });
        }
    };
    /**
     * @private
     * @return {?}
     */
    FormatlistComponent.prototype.checkBeforePrint = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.formatId) {
            this.notifyService.warning(this.localepip.transform('selectformat'));
            return false;
        }
        else {
            return true;
        }
    };
    FormatlistComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-formatlist',
                    template: "\r\n<div class=\"f-page f-page-navigate f-page-is-grid-card\">\r\n    <section class=\"f-page-main\">\r\n        <farris-section class=\"f-section-grid f-section-in-managelist\" [showHeader]=\"false\">\r\n    \r\n            <div class=\"f-section-content\">\r\n                <div class=\"f-component-grid\">\r\n                <farris-datagrid\r\n                    [fit]=\"true\" \r\n                    [pagination]=\"false\"\r\n                    [virtualized]=\"false\"\r\n                    [columns]=\"columns\"\r\n                    [pageSize]=\"50\"\r\n                    [showBorder]=\"false\"\r\n                    [striped]=\"false\"\r\n                    [data]=\"formats\"\r\n                    [fitColumns]=\"true\"\r\n                    [nowrap]=\"false\"\r\n                    (selectChanged)=\"selectChange($event)\">\r\n                </farris-datagrid>\r\n                </div>\r\n            </div>\r\n        </farris-section>\r\n    </section>    \r\n</div>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    FormatlistComponent.ctorParameters = function () { return [
        { type: NotifyService },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    FormatlistComponent.propDecorators = {
        formats: [{ type: Input }]
    };
    return FormatlistComponent;
}());
export { FormatlistComponent };
if (false) {
    /** @type {?} */
    FormatlistComponent.prototype.formats;
    /** @type {?} */
    FormatlistComponent.prototype.print;
    /** @type {?} */
    FormatlistComponent.prototype.columns;
    /** @type {?} */
    FormatlistComponent.prototype.formatId;
    /** @type {?} */
    FormatlistComponent.prototype.localepip;
    /**
     * @type {?}
     * @private
     */
    FormatlistComponent.prototype.notifyService;
}
