// @farris/ui-html-editor/esm2015/lib/html-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, forwardRef, Inject, Input, NgZone, Output, PLATFORM_ID, Renderer2, SecurityContext, ViewEncapsulation, HostBinding, Optional, Injector } from '@angular/core';
import { isPlatformServer, DOCUMENT } from '@angular/common';
import { DomSanitizer } from '@angular/platform-browser';
import { NG_VALUE_ACCESSOR, NG_VALIDATORS } from '@angular/forms';
import Quill from 'quill';
import { Convert2HtmlEditorToolbars } from '@farris/ui-common';
import { defaultModules } from './html-editor-defaults';
import { QUILL_CONFIG_TOKEN } from './html-editor.interfaces';
import { ImageResize } from './image-resize-module/ImageResize';
import { ImageDrop } from './image-drop-module/image-drop'; // 拖动加载图片组件。
// 拖动加载图片组件。
import { Tooltip } from './tool-tip-module/index';
import { NotifyService } from '@farris/ui-notify';
import { LocaleService } from '@farris/ui-locale';
import { HTML_EDITOR_LOCALES } from './html-editor.locale';
// 修改默认字体
/** @type {?} */
const Font = Quill.import('attributors/style/font');
Font.whitelist = ['Microsoft-YaHei', 'SimSun', 'SimHei', 'KaiTi', 'FangSong', 'Arial', 'Times', 'sans-serif'];
Quill.register(Font, true);
// 注册图片拖拽和图片大小调整
Quill.register('modules/imageDrop', ImageDrop);
Quill.register('modules/imageResize', ImageResize);
Quill.register('modules/tooltip', Tooltip);
/**
 * @record
 */
export function CustomOption() { }
if (false) {
    /** @type {?} */
    CustomOption.prototype.import;
    /** @type {?} */
    CustomOption.prototype.whitelist;
}
/**
 * @record
 */
export function Range() { }
if (false) {
    /** @type {?} */
    Range.prototype.index;
    /** @type {?} */
    Range.prototype.length;
}
/** @type {?} */
const getFormat = (/**
 * @param {?=} format
 * @param {?=} configFormat
 * @return {?}
 */
(format, configFormat) => {
    /** @type {?} */
    const passedFormat = format || configFormat;
    return passedFormat || 'html';
});
const ɵ0 = getFormat;
export class HtmlEditorComponent {
    /**
     * @param {?} elementRef
     * @param {?} domSanitizer
     * @param {?} doc
     * @param {?} platformId
     * @param {?} renderer
     * @param {?} zone
     * @param {?} config
     * @param {?} notifySer
     * @param {?} localeSer
     * @param {?} injector
     */
    constructor(elementRef, domSanitizer, doc, platformId, renderer, zone, config, notifySer, localeSer, injector) {
        this.elementRef = elementRef;
        this.domSanitizer = domSanitizer;
        this.doc = doc;
        this.platformId = platformId;
        this.renderer = renderer;
        this.zone = zone;
        this.config = config;
        this.notifySer = notifySer;
        this.localeSer = localeSer;
        this.injector = injector;
        this._disabled = false; // used to store initial value before ViewInit
        // used to store initial value before ViewInit
        this.rootClass = 'farris-html-editor';
        this.disabled = false;
        this.showBorder = true;
        this.modules = {
            imageDrop: true,
            imageResize: true,
            tooltip: true
        };
        this.required = false;
        this.customToolbarPosition = 'top';
        this.sanitize = false; // 净化dom
        // 净化dom
        this.styles = null;
        this.strict = true;
        this.customOptions = [];
        this.preserveWhitespace = false;
        /**
         * 限制上传图片大小, 默认为 false不限制上传图片的大小
         */
        this.limitImageSize = false;
        /**
         * 允许上传图片最大字节数，单位 kb
         */
        this.maxImgSize = 500;
        // 自定义toobar
        this.customToolbar = false;
        // 文字样式
        this.fontStyleBar = true;
        // 代码块
        this.BlockBar = false; // false;
        // false;
        // 标题
        this.headerBar = true;
        // 字号
        this.sizeBar = true;
        // 列表
        this.listBar = true;
        // 字体
        this.fontBar = true;
        // 上标和下标
        this.superAndSubscriptBar = true; // false;
        // false;
        // 缩进
        this.indentBar = true;
        // 文字颜色
        this.colorBar = true; // false;
        // false;
        // 媒体
        this.mediaBar = true; // false;
        // false;
        // 图片
        this.imageBar = false;
        // 视频
        this.videoBar = false; // false;
        // false;
        // 清除
        this.cleanBar = true;
        this.alignBar = true;
        /**
         * 只读禁用时，最大高度，为 0 时，不限制最大高度; 默认 800px
         */
        this.maxHeight = 800;
        this.minHeight = 100;
        this.autoHeight = true;
        this.onEditorCreated = new EventEmitter();
        this.onContentChanged = new EventEmitter();
        this.onSelectionChanged = new EventEmitter();
        this.onFocus = new EventEmitter();
        this.onBlur = new EventEmitter();
        this.valueGetter = (/**
         * @param {?} quillEditor
         * @param {?} editorElement
         * @return {?}
         */
        (quillEditor, editorElement) => {
            // tslint:disable-next-line: no-non-null-assertion
            /** @type {?} */
            let html = (/** @type {?} */ (editorElement.querySelector('.ql-editor'))).innerHTML;
            if (html === '<p><br></p>' || html === '<div><br><div>') {
                html = null;
            }
            /** @type {?} */
            let modelValue = html;
            /** @type {?} */
            const format = getFormat(this.format, this.config.format);
            if (format === 'text') {
                modelValue = quillEditor.getText();
            }
            else if (format === 'object') {
                modelValue = quillEditor.getContents();
            }
            else if (format === 'json') {
                try {
                    modelValue = JSON.stringify(quillEditor.getContents());
                }
                catch (e) {
                    modelValue = quillEditor.getText();
                }
            }
            return modelValue;
        });
        this.valueSetter = (/**
         * @param {?} quillEditor
         * @param {?} value
         * @return {?}
         */
        (quillEditor, value) => {
            /** @type {?} */
            const format = getFormat(this.format, this.config.format);
            if (format === 'html') {
                if (this.sanitize) {
                    value = this.domSanitizer.sanitize(SecurityContext.HTML, value);
                }
                value = value ? (value + '').replace(/\s\s/g, '&nbsp;&nbsp;') : '';
                return quillEditor.clipboard.convert(value);
            }
            else if (format === 'json') {
                try {
                    return JSON.parse(value);
                }
                catch (e) {
                    return [{ insert: value }];
                }
            }
            return value;
        });
        this.selectionChangeHandler = (/**
         * @param {?} range
         * @param {?} oldRange
         * @param {?} source
         * @return {?}
         */
        (range, oldRange, source) => {
            this.zone.run((/**
             * @return {?}
             */
            () => {
                if (range === null) {
                    this.onBlur.emit({
                        editor: this.quillEditor,
                        source
                    });
                }
                else if (oldRange === null) {
                    this.onFocus.emit({
                        editor: this.quillEditor,
                        source
                    });
                }
                this.onSelectionChanged.emit({
                    editor: this.quillEditor,
                    oldRange,
                    range,
                    source
                });
                if (!range && this.onModelTouched) {
                    this.onModelTouched();
                }
            }));
        });
        this.textChangeHandler = (/**
         * @param {?} delta
         * @param {?} oldDelta
         * @param {?} source
         * @return {?}
         */
        (delta, oldDelta, source) => {
            // only emit changes emitted by user interactions
            // only emit changes emitted by user interactions
            /** @type {?} */
            const text = this.quillEditor.getText();
            /** @type {?} */
            const content = this.quillEditor.getContents();
            // tslint:disable-next-line: no-non-null-assertion
            /** @type {?} */
            let html = (/** @type {?} */ ((/** @type {?} */ (this.editorElem)).querySelector('.ql-editor'))).innerHTML;
            if (html === '<p><br></p>' || html === '<div><br><div>') {
                html = null;
            }
            this.zone.run((/**
             * @return {?}
             */
            () => {
                /** @type {?} */
                const trackChanges = this.trackChanges || this.config.trackChanges;
                if ((source === Quill.sources.USER || (trackChanges && trackChanges === 'all')) && this.onModelChange) {
                    this.onModelChange(this.valueGetter(this.quillEditor, (/** @type {?} */ (this.editorElem))));
                }
                this.onContentChanged.emit({
                    content,
                    delta,
                    editor: this.quillEditor,
                    html,
                    oldDelta,
                    source,
                    text
                });
            }));
        });
        if (this.localeSer) {
            this.localeSer.setLocaleData(HTML_EDITOR_LOCALES);
        }
    }
    // tslint:disable-next-line:no-empty
    /**
     * @param {?=} _modelValue
     * @return {?}
     */
    onModelChange(_modelValue) { }
    // tslint:disable-next-line:no-empty
    /**
     * @return {?}
     */
    onModelTouched() { }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.editorCreate();
        if (this.editorElem) {
            this.renderer.setStyle(this.editorElem, 'min-height', `${this.minHeight || 100}px`);
            if (this.autoHeight) {
                if (this.maxHeight) {
                    this.renderer.setStyle(this.editorElem, 'max-height', `${this.maxHeight}px`);
                    this.renderer.setStyle(this.editorElem, 'overflow', 'auto');
                }
            }
            else {
                this.renderer.setStyle(this.editorElem, 'max-height', `${this.minHeight}px`);
                this.renderer.setStyle(this.editorElem, 'overflow', 'auto');
            }
        }
    }
    /**
     * @param {?} toolbar
     * @return {?}
     */
    setToolbar(toolbar) {
        // console.log(toolbar);
    }
    /**
     * @private
     * @return {?}
     */
    editorCreate() {
        if (isPlatformServer(this.platformId)) {
            return;
        }
        this.elementRef.nativeElement.insertAdjacentHTML(this.customToolbarPosition === 'top' ? 'beforeend' : 'afterbegin', this.preserveWhitespace ? '<pre html-editor-element></pre>' : '<div html-editor-element></div>');
        this.editorElem = this.elementRef.nativeElement.querySelector('[html-editor-element]');
        /** @type {?} */
        const toolbarElem = this.elementRef.nativeElement.querySelector('[html-editor-toolbar]');
        /** @type {?} */
        const modules = this.modules || this.config.modules || defaultModules;
        if (modules.toolbar === undefined) {
            modules.toolbar = defaultModules.toolbar;
        }
        if (this.customToolbar) {
            if (this.toolbar && this.toolbar.length) {
                modules.toolbar = Convert2HtmlEditorToolbars('concise', this.toolbar);
            }
            else {
                modules.toolbar = defaultModules.toolbar;
            }
        }
        if (modules.tooltip === undefined) {
            modules.tooltip = defaultModules.tooltip;
        }
        if (modules.imageDrop === undefined) {
            modules.imageDrop = defaultModules.imageDrop;
        }
        if (modules.imageResize === undefined) {
            modules.imageResize = defaultModules.imageResize;
        }
        /** @type {?} */
        let placeholder = this.placeholder !== undefined ? this.placeholder : this.config.placeholder;
        if (placeholder === undefined) {
            placeholder = '';
        }
        if (toolbarElem) {
            // tslint:disable-next-line:no-string-literal
            modules['toolbar'] = toolbarElem;
        }
        if (this.styles) {
            Object.keys(this.styles).forEach((/**
             * @param {?} key
             * @return {?}
             */
            (key) => {
                this.renderer.setStyle(this.editorElem, key, this.styles[key]);
            }));
        }
        this.customOptions.forEach((/**
         * @param {?} customOption
         * @return {?}
         */
        customOption => {
            /** @type {?} */
            const newCustomOption = Quill.import(customOption.import);
            newCustomOption.whitelist = customOption.whitelist;
            Quill.register(newCustomOption, true);
        }));
        /** @type {?} */
        let bounds = this.bounds && this.bounds === 'self' ? this.editorElem : this.bounds;
        if (!bounds) {
            bounds = this.config.bounds ? this.config.bounds : this.doc.body;
        }
        /** @type {?} */
        let debug = this.debug;
        if (!debug && debug !== false && this.config.debug) {
            debug = this.config.debug;
        }
        /** @type {?} */
        let readOnly = this.readonly;
        if (!readOnly && this.readonly !== false) {
            readOnly = this.config.readOnly !== undefined ? this.config.readOnly : false;
        }
        // modules.imageResize = !(readOnly || this.disabled)
        /** @type {?} */
        let scrollingContainer = this.scrollingContainer;
        if (!scrollingContainer) {
            scrollingContainer = this.editorElem;
        }
        // if (!scrollingContainer) {
        //     scrollingContainer =
        //         this.config.scrollingContainer === null || this.config.scrollingContainer
        //             ? this.config.scrollingContainer
        //             : null;
        // }
        /** @type {?} */
        let formats = this.formats;
        if (!formats && formats === undefined) {
            formats = this.config.formats || this.config.formats === null ? this.config.formats : undefined;
        }
        // 初始化quill
        this.quillEditor = new Quill(this.editorElem, {
            bounds,
            debug,
            formats,
            modules,
            placeholder,
            readOnly,
            scrollingContainer,
            strict: this.strict,
            theme: this.theme || (this.config.theme ? this.config.theme : 'snow'),
            /**
             * 限制上传图片大小, 默认为 false不限制上传图片的大小
             */
            limitImageSize: this.limitImageSize,
            /**
             * 允许上传图片最大字节数，单位 kb
             */
            maxImgSize: this.maxImgSize,
            showMaxImageMessage: this.showMaxImageMessage.bind(this),
            ngZone: this.zone,
            localeId: this.localeSer ? this.localeSer.localeId : localStorage.getItem('languageCode') || 'zh-CHS'
        });
        if (this.content) {
            /** @type {?} */
            const format = getFormat(this.format, this.config.format);
            if (format === 'object') {
                this.quillEditor.setContents(this.content, 'silent');
            }
            else if (format === 'text') {
                this.quillEditor.setText(this.content, 'silent');
            }
            else if (format === 'json') {
                try {
                    this.quillEditor.setContents(JSON.parse(this.content), 'silent');
                }
                catch (e) {
                    this.quillEditor.setText(this.content, 'silent');
                }
            }
            else {
                if (this.sanitize) {
                    this.content = this.domSanitizer.sanitize(SecurityContext.HTML, this.content);
                }
                /** @type {?} */
                const contents = this.quillEditor.clipboard.convert(this.content);
                this.quillEditor.setContents(contents, 'silent');
            }
            this.quillEditor.history.clear();
        }
        // initialize _disabled status based on this._disabled as default value
        // this.setDisabledState(this.disabled);
        this.setStatusEditor(this.readonly, this.disabled);
        this.onEditorCreated.emit(this.quillEditor);
        // mark model as touched if editor lost focus
        this.quillEditor.on('selection-change', this.selectionChangeHandler);
        // update model if text changes
        this.quillEditor.on('text-change', this.textChangeHandler);
    }
    /**
     * @return {?}
     */
    random_id() {
        return Math.random().toString(36).slice(2);
    }
    /**
     * @return {?}
     */
    endEdit() {
        this.onModelTouched();
    }
    /**
     * @param {?} options
     * @return {?}
     */
    buildToolbar(options) {
        /** @type {?} */
        const rtn = [];
        Object.keys(options).forEach((/**
         * @param {?} e
         * @return {?}
         */
        e => {
            if (this[e] === true) {
                rtn.push(options[e]);
            }
            else if (this[e]) {
                rtn.push(this[e]);
            }
        }));
        return rtn;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.quillEditor) {
            this.quillEditor.off('selection-change', this.selectionChangeHandler);
            this.quillEditor.off('text-change', this.textChangeHandler);
        }
    }
    /**
     * @private
     * @return {?}
     */
    hideImgSelectionRange() {
        /** @type {?} */
        const imgResizeIns = this.quillEditor.getModule('imageResize');
        if (imgResizeIns) {
            if (this.readonly || this.disabled) {
                imgResizeIns.hide();
            }
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (!this.quillEditor) {
            return;
        }
        if (changes['disabled']) {
            this.setStatusEditor(false, this.disabled);
            this.hideImgSelectionRange();
        }
        // tslint:disable:no-string-literal
        if (changes['readonly']) {
            this.quillEditor.enable(!changes['readonly'].currentValue);
            this.disableToolbar(changes['readonly'].currentValue);
            this.hideImgSelectionRange();
        }
        if (changes['placeholder']) {
            this.quillEditor.root.dataset.placeholder = changes['placeholder'].currentValue;
        }
        if (changes['styles']) {
            /** @type {?} */
            const currentStyling = changes['styles'].currentValue;
            /** @type {?} */
            const previousStyling = changes['styles'].previousValue;
            if (previousStyling) {
                Object.keys(previousStyling).forEach((/**
                 * @param {?} key
                 * @return {?}
                 */
                (key) => {
                    this.renderer.removeStyle(this.editorElem, key);
                }));
            }
            if (currentStyling) {
                Object.keys(currentStyling).forEach((/**
                 * @param {?} key
                 * @return {?}
                 */
                (key) => {
                    this.renderer.setStyle(this.editorElem, key, this.styles[key]);
                }));
            }
        }
        if (changes.limitImageSize && !changes.limitImageSize.isFirstChange()) {
            this.quillEditor.options.limitImageSize = this.limitImageSize;
        }
        if (changes.maxImgSize && !changes.maxImgSize.isFirstChange()) {
            this.quillEditor.options.maxImgSize = this.maxImgSize;
        }
        // tslint:enable:no-string-literal
    }
    /**
     * @param {?} currentValue
     * @return {?}
     */
    writeValue(currentValue) {
        this.content = currentValue;
        /** @type {?} */
        const format = getFormat(this.format, this.config.format);
        if (this.quillEditor) {
            this.quillEditor.setText('');
            if (currentValue) {
                if (format === 'text') {
                    this.quillEditor.setText(currentValue);
                }
                else {
                    this.quillEditor.setContents(this.valueSetter(this.quillEditor, this.content));
                }
                return;
            }
        }
    }
    /**
     * @param {?} readonly
     * @return {?}
     */
    disableToolbar(readonly) {
        /** @type {?} */
        const toolbarElem = this.elementRef.nativeElement.querySelector('.ql-toolbar');
        if (toolbarElem) {
            toolbarElem.style.display = readonly ? 'none' : 'block';
        }
    }
    /**
     * @param {?=} readonly
     * @param {?=} disable
     * @return {?}
     */
    setStatusEditor(readonly = false, disable) {
        // if (disable || readonly) {
        //     this.elementRef.nativeElement.style.border = 'none';
        // }
        this.disableToolbar(disable || readonly);
        this.setDisabledState(disable);
    }
    /**
     * @param {?=} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled = this._disabled) {
        // store initial value to set appropriate _disabled status after ViewInit
        this._disabled = isDisabled;
        if (this.quillEditor) {
            if (isDisabled) {
                this.quillEditor.disable();
                this.renderer.setAttribute(this.elementRef.nativeElement, 'disabled', 'disabled');
            }
            else {
                if (!this.readonly) {
                    this.quillEditor.enable();
                }
                this.renderer.removeAttribute(this.elementRef.nativeElement, 'disabled');
            }
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onModelChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onModelTouched = fn;
    }
    /**
     * @param {?=} imgs
     * @return {?}
     */
    showMaxImageMessage(imgs = null) {
        /** @type {?} */
        let errMsg = `上传的图片不得大于${this.maxImgSize}kb`;
        if (this.localeSer) {
            /** @type {?} */
            let _errMsg = this.localeSer.getValue('html-editor.maximgsize');
            if (_errMsg) {
                _errMsg = _errMsg.replace('{0}', `${this.maxImgSize}kb`);
                errMsg = _errMsg;
            }
            if (imgs && imgs.length) {
                /** @type {?} */
                const imgNames = imgs.map((/**
                 * @param {?} n
                 * @return {?}
                 */
                n => n.name)).join(', ');
                errMsg = this.localeSer.getValue('html-editor.dropimg').replace('{0}', imgNames).replace('{1}', `${this.maxImgSize}kb`);
            }
        }
        else {
            if (imgs && imgs.length) {
                /** @type {?} */
                const imgNames = imgs.map((/**
                 * @param {?} n
                 * @return {?}
                 */
                n => n.name)).join(', ');
                errMsg = `${imgNames} 超出上传限制！最大允许 ${this.maxImgSize}kb`;
            }
        }
        if (this.notifySer) {
            this.notifySer.warning({ msg: errMsg, timeout: 2000 });
        }
        else {
            alert(errMsg);
        }
    }
    /**
     * @return {?}
     */
    validate() {
        return null;
        // if (!this.quillEditor) {
        // }
        // const err: {
        //     minLengthError?: {
        //         given: number;
        //         minLength: number;
        //     };
        //     maxLengthError?: {
        //         given: number;
        //         maxLength: number;
        //     };
        //     requiredError?: { empty: boolean };
        // } = {};
        // let valid = true;
        // const textLength = this.quillEditor.getText().trim().length;
        // if (this.minLength && textLength && textLength < this.minLength) {
        //     err.minLengthError = {
        //         given: textLength,
        //         minLength: this.minLength
        //     };
        //     valid = false;
        // }
        // if (this.maxLength && textLength > this.maxLength) {
        //     err.maxLengthError = {
        //         given: textLength,
        //         maxLength: this.maxLength
        //     };
        //     valid = false;
        // }
        // if (this.required && !textLength) {
        //     err.requiredError = {
        //         empty: true
        //     };
        //     valid = false;
        // }
        // return valid ? null : err;
    }
}
HtmlEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'farris-html-editor',
                template: `
        <ng-content select="[html-editor-toolbar]"></ng-content>
    `,
                encapsulation: ViewEncapsulation.None,
                providers: [
                    {
                        multi: true,
                        provide: NG_VALUE_ACCESSOR,
                        // 自定义组件的数据绑定
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => HtmlEditorComponent))
                    }, {
                        provide: NG_VALIDATORS,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => HtmlEditorComponent)),
                        multi: true
                    }
                ],
                styles: [".ql-container{box-sizing:border-box;font-family:Helvetica,Arial,sans-serif;font-size:.875rem;height:100%;margin:0;position:relative;color:rgba(0,0,0,.85)}.ql-container.ql-disabled .ql-tooltip{visibility:hidden}.ql-container.ql-disabled .ql-editor ul[data-checked]>li::before{pointer-events:none}.ql-clipboard{left:-100000px;height:1px;overflow-y:hidden;position:absolute;top:50%}.ql-clipboard p{margin:0;padding:0}.ql-editor{box-sizing:border-box;line-height:1.42;height:100%;outline:0;overflow-y:auto;padding:12px 15px;-o-tab-size:4;tab-size:4;-moz-tab-size:4;text-align:left;white-space:pre-wrap;word-wrap:break-word;min-height:80px;word-break:break-all}.ql-editor>*{cursor:text}.ql-editor blockquote,.ql-editor h1,.ql-editor h2,.ql-editor h3,.ql-editor h4,.ql-editor h5,.ql-editor h6,.ql-editor ol,.ql-editor p,.ql-editor pre,.ql-editor ul{margin:0;padding:0;counter-reset:list-1 list-2 list-3 list-4 list-5 list-6 list-7 list-8 list-9}.ql-editor ol,.ql-editor ul{padding-left:1.5em}.ql-editor ol>li,.ql-editor ul>li{list-style-type:none}.ql-editor ul>li::before{content:'\\2022'}.ql-editor ul[data-checked=false],.ql-editor ul[data-checked=true]{pointer-events:none}.ql-editor ul[data-checked=false]>li *,.ql-editor ul[data-checked=true]>li *{pointer-events:all}.ql-editor ul[data-checked=false]>li::before,.ql-editor ul[data-checked=true]>li::before{color:#777;cursor:pointer;pointer-events:all}.ql-editor ul[data-checked=true]>li::before{content:'\\2611'}.ql-editor ul[data-checked=false]>li::before{content:'\\2610'}.ql-editor li::before{display:inline-block;white-space:nowrap;width:1.2em}.ql-editor li:not(.ql-direction-rtl)::before{margin-left:-1.5em;margin-right:.3em;text-align:right}.ql-editor li.ql-direction-rtl::before{margin-left:.3em;margin-right:-1.5em}.ql-editor ol li:not(.ql-direction-rtl),.ql-editor ul li:not(.ql-direction-rtl){padding-left:1.5em}.ql-editor ol li.ql-direction-rtl,.ql-editor ul li.ql-direction-rtl{padding-right:1.5em}.ql-editor ol li{counter-reset:list-1 list-2 list-3 list-4 list-5 list-6 list-7 list-8 list-9;counter-increment:list-0}.ql-editor ol li:before{content:counter(list-0,decimal) '. '}.ql-editor ol li.ql-indent-1{counter-increment:list-1;counter-reset:list-2 list-3 list-4 list-5 list-6 list-7 list-8 list-9}.ql-editor ol li.ql-indent-1:before{content:counter(list-1,lower-alpha) '. '}.ql-editor ol li.ql-indent-2{counter-increment:list-2;counter-reset:list-3 list-4 list-5 list-6 list-7 list-8 list-9}.ql-editor ol li.ql-indent-2:before{content:counter(list-2,lower-roman) '. '}.ql-editor ol li.ql-indent-3{counter-increment:list-3;counter-reset:list-4 list-5 list-6 list-7 list-8 list-9}.ql-editor ol li.ql-indent-3:before{content:counter(list-3,decimal) '. '}.ql-editor ol li.ql-indent-4{counter-increment:list-4;counter-reset:list-5 list-6 list-7 list-8 list-9}.ql-editor ol li.ql-indent-4:before{content:counter(list-4,lower-alpha) '. '}.ql-editor ol li.ql-indent-5{counter-increment:list-5;counter-reset:list-6 list-7 list-8 list-9}.ql-editor ol li.ql-indent-5:before{content:counter(list-5,lower-roman) '. '}.ql-editor ol li.ql-indent-6{counter-increment:list-6;counter-reset:list-7 list-8 list-9}.ql-editor ol li.ql-indent-6:before{content:counter(list-6,decimal) '. '}.ql-editor ol li.ql-indent-7{counter-increment:list-7;counter-reset:list-8 list-9}.ql-editor ol li.ql-indent-7:before{content:counter(list-7,lower-alpha) '. '}.ql-editor ol li.ql-indent-8{counter-increment:list-8;counter-reset:list-9}.ql-editor ol li.ql-indent-8:before{content:counter(list-8,lower-roman) '. '}.ql-editor ol li.ql-indent-9{counter-increment:list-9}.ql-editor ol li.ql-indent-9:before{content:counter(list-9,decimal) '. '}.ql-editor .ql-indent-1:not(.ql-direction-rtl){padding-left:3em}.ql-editor li.ql-indent-1:not(.ql-direction-rtl){padding-left:4.5em}.ql-editor .ql-indent-1.ql-direction-rtl.ql-align-right{padding-right:3em}.ql-editor li.ql-indent-1.ql-direction-rtl.ql-align-right{padding-right:4.5em}.ql-editor .ql-indent-2:not(.ql-direction-rtl){padding-left:6em}.ql-editor li.ql-indent-2:not(.ql-direction-rtl){padding-left:7.5em}.ql-editor .ql-indent-2.ql-direction-rtl.ql-align-right{padding-right:6em}.ql-editor li.ql-indent-2.ql-direction-rtl.ql-align-right{padding-right:7.5em}.ql-editor .ql-indent-3:not(.ql-direction-rtl){padding-left:9em}.ql-editor li.ql-indent-3:not(.ql-direction-rtl){padding-left:10.5em}.ql-editor .ql-indent-3.ql-direction-rtl.ql-align-right{padding-right:9em}.ql-editor li.ql-indent-3.ql-direction-rtl.ql-align-right{padding-right:10.5em}.ql-editor .ql-indent-4:not(.ql-direction-rtl){padding-left:12em}.ql-editor li.ql-indent-4:not(.ql-direction-rtl){padding-left:13.5em}.ql-editor .ql-indent-4.ql-direction-rtl.ql-align-right{padding-right:12em}.ql-editor li.ql-indent-4.ql-direction-rtl.ql-align-right{padding-right:13.5em}.ql-editor .ql-indent-5:not(.ql-direction-rtl){padding-left:15em}.ql-editor li.ql-indent-5:not(.ql-direction-rtl){padding-left:16.5em}.ql-editor .ql-indent-5.ql-direction-rtl.ql-align-right{padding-right:15em}.ql-editor li.ql-indent-5.ql-direction-rtl.ql-align-right{padding-right:16.5em}.ql-editor .ql-indent-6:not(.ql-direction-rtl){padding-left:18em}.ql-editor li.ql-indent-6:not(.ql-direction-rtl){padding-left:19.5em}.ql-editor .ql-indent-6.ql-direction-rtl.ql-align-right{padding-right:18em}.ql-editor li.ql-indent-6.ql-direction-rtl.ql-align-right{padding-right:19.5em}.ql-editor .ql-indent-7:not(.ql-direction-rtl){padding-left:21em}.ql-editor li.ql-indent-7:not(.ql-direction-rtl){padding-left:22.5em}.ql-editor .ql-indent-7.ql-direction-rtl.ql-align-right{padding-right:21em}.ql-editor li.ql-indent-7.ql-direction-rtl.ql-align-right{padding-right:22.5em}.ql-editor .ql-indent-8:not(.ql-direction-rtl){padding-left:24em}.ql-editor li.ql-indent-8:not(.ql-direction-rtl){padding-left:25.5em}.ql-editor .ql-indent-8.ql-direction-rtl.ql-align-right{padding-right:24em}.ql-editor li.ql-indent-8.ql-direction-rtl.ql-align-right{padding-right:25.5em}.ql-editor .ql-indent-9:not(.ql-direction-rtl){padding-left:27em}.ql-editor li.ql-indent-9:not(.ql-direction-rtl){padding-left:28.5em}.ql-editor .ql-indent-9.ql-direction-rtl.ql-align-right{padding-right:27em}.ql-editor li.ql-indent-9.ql-direction-rtl.ql-align-right{padding-right:28.5em}.ql-editor .ql-video{display:block;max-width:100%}.ql-editor .ql-video.ql-align-center{margin:0 auto}.ql-editor .ql-video.ql-align-right{margin:0 0 0 auto}.ql-editor .ql-bg-black{background-color:#000}.ql-editor .ql-bg-red{background-color:#e60000}.ql-editor .ql-bg-orange{background-color:#f90}.ql-editor .ql-bg-yellow{background-color:#ff0}.ql-editor .ql-bg-green{background-color:#008a00}.ql-editor .ql-bg-blue{background-color:#06c}.ql-editor .ql-bg-purple{background-color:#93f}.ql-editor .ql-color-white{color:#fff}.ql-editor .ql-color-red{color:#e60000}.ql-editor .ql-color-orange{color:#f90}.ql-editor .ql-color-yellow{color:#ff0}.ql-editor .ql-color-green{color:#008a00}.ql-editor .ql-color-blue{color:#06c}.ql-editor .ql-color-purple{color:#93f}.ql-editor .ql-font-serif{font-family:Georgia,Times New Roman,serif}.ql-editor .ql-font-monospace{font-family:Monaco,Courier New,monospace}.ql-editor .ql-size-small{font-size:.75em}.ql-editor .ql-size-large{font-size:1.5em}.ql-editor .ql-size-huge{font-size:2.5em}.ql-editor .ql-direction-rtl{direction:rtl;text-align:inherit}.ql-editor .ql-align-center{text-align:center}.ql-editor .ql-align-justify{text-align:justify}.ql-editor .ql-align-right{text-align:right}.ql-editor.ql-blank::before{color:rgba(0,0,0,.6);content:attr(data-placeholder);left:15px;pointer-events:none;position:absolute;right:15px}.ql-snow .ql-toolbar:after,.ql-snow.ql-toolbar:after{clear:both;content:'';display:table}.ql-snow .ql-toolbar button,.ql-snow.ql-toolbar button{background:0 0;border:none;cursor:pointer;display:inline-block;float:left;height:24px;padding:3px 5px;width:28px}.ql-snow .ql-toolbar button svg,.ql-snow.ql-toolbar button svg{float:left;height:100%}.ql-snow .ql-toolbar button:active:hover,.ql-snow.ql-toolbar button:active:hover{outline:0}.ql-snow .ql-toolbar input.ql-image[type=file],.ql-snow.ql-toolbar input.ql-image[type=file]{display:none}.ql-snow .ql-toolbar .ql-picker-item.ql-selected,.ql-snow .ql-toolbar .ql-picker-item:hover,.ql-snow .ql-toolbar .ql-picker-label.ql-active,.ql-snow .ql-toolbar .ql-picker-label:hover,.ql-snow .ql-toolbar button.ql-active,.ql-snow .ql-toolbar button:focus,.ql-snow .ql-toolbar button:hover,.ql-snow.ql-toolbar .ql-picker-item.ql-selected,.ql-snow.ql-toolbar .ql-picker-item:hover,.ql-snow.ql-toolbar .ql-picker-label.ql-active,.ql-snow.ql-toolbar .ql-picker-label:hover,.ql-snow.ql-toolbar button.ql-active,.ql-snow.ql-toolbar button:focus,.ql-snow.ql-toolbar button:hover{color:#06c}.ql-snow .ql-toolbar .ql-picker-item.ql-selected .ql-fill,.ql-snow .ql-toolbar .ql-picker-item.ql-selected .ql-stroke.ql-fill,.ql-snow .ql-toolbar .ql-picker-item:hover .ql-fill,.ql-snow .ql-toolbar .ql-picker-item:hover .ql-stroke.ql-fill,.ql-snow .ql-toolbar .ql-picker-label.ql-active .ql-fill,.ql-snow .ql-toolbar .ql-picker-label.ql-active .ql-stroke.ql-fill,.ql-snow .ql-toolbar .ql-picker-label:hover .ql-fill,.ql-snow .ql-toolbar .ql-picker-label:hover .ql-stroke.ql-fill,.ql-snow .ql-toolbar button.ql-active .ql-fill,.ql-snow .ql-toolbar button.ql-active .ql-stroke.ql-fill,.ql-snow .ql-toolbar button:focus .ql-fill,.ql-snow .ql-toolbar button:focus .ql-stroke.ql-fill,.ql-snow .ql-toolbar button:hover .ql-fill,.ql-snow .ql-toolbar button:hover .ql-stroke.ql-fill,.ql-snow.ql-toolbar .ql-picker-item.ql-selected .ql-fill,.ql-snow.ql-toolbar .ql-picker-item.ql-selected .ql-stroke.ql-fill,.ql-snow.ql-toolbar .ql-picker-item:hover .ql-fill,.ql-snow.ql-toolbar .ql-picker-item:hover .ql-stroke.ql-fill,.ql-snow.ql-toolbar .ql-picker-label.ql-active .ql-fill,.ql-snow.ql-toolbar .ql-picker-label.ql-active .ql-stroke.ql-fill,.ql-snow.ql-toolbar .ql-picker-label:hover .ql-fill,.ql-snow.ql-toolbar .ql-picker-label:hover .ql-stroke.ql-fill,.ql-snow.ql-toolbar button.ql-active .ql-fill,.ql-snow.ql-toolbar button.ql-active .ql-stroke.ql-fill,.ql-snow.ql-toolbar button:focus .ql-fill,.ql-snow.ql-toolbar button:focus .ql-stroke.ql-fill,.ql-snow.ql-toolbar button:hover .ql-fill,.ql-snow.ql-toolbar button:hover .ql-stroke.ql-fill{fill:#06c}.ql-snow .ql-toolbar .ql-picker-item.ql-selected .ql-stroke,.ql-snow .ql-toolbar .ql-picker-item.ql-selected .ql-stroke-miter,.ql-snow .ql-toolbar .ql-picker-item:hover .ql-stroke,.ql-snow .ql-toolbar .ql-picker-item:hover .ql-stroke-miter,.ql-snow .ql-toolbar .ql-picker-label.ql-active .ql-stroke,.ql-snow .ql-toolbar .ql-picker-label.ql-active .ql-stroke-miter,.ql-snow .ql-toolbar .ql-picker-label:hover .ql-stroke,.ql-snow .ql-toolbar .ql-picker-label:hover .ql-stroke-miter,.ql-snow .ql-toolbar button.ql-active .ql-stroke,.ql-snow .ql-toolbar button.ql-active .ql-stroke-miter,.ql-snow .ql-toolbar button:focus .ql-stroke,.ql-snow .ql-toolbar button:focus .ql-stroke-miter,.ql-snow .ql-toolbar button:hover .ql-stroke,.ql-snow .ql-toolbar button:hover .ql-stroke-miter,.ql-snow.ql-toolbar .ql-picker-item.ql-selected .ql-stroke,.ql-snow.ql-toolbar .ql-picker-item.ql-selected .ql-stroke-miter,.ql-snow.ql-toolbar .ql-picker-item:hover .ql-stroke,.ql-snow.ql-toolbar .ql-picker-item:hover .ql-stroke-miter,.ql-snow.ql-toolbar .ql-picker-label.ql-active .ql-stroke,.ql-snow.ql-toolbar .ql-picker-label.ql-active .ql-stroke-miter,.ql-snow.ql-toolbar .ql-picker-label:hover .ql-stroke,.ql-snow.ql-toolbar .ql-picker-label:hover .ql-stroke-miter,.ql-snow.ql-toolbar button.ql-active .ql-stroke,.ql-snow.ql-toolbar button.ql-active .ql-stroke-miter,.ql-snow.ql-toolbar button:focus .ql-stroke,.ql-snow.ql-toolbar button:focus .ql-stroke-miter,.ql-snow.ql-toolbar button:hover .ql-stroke,.ql-snow.ql-toolbar button:hover .ql-stroke-miter{stroke:#06c}@media (pointer:coarse){.ql-snow .ql-toolbar button:hover:not(.ql-active),.ql-snow.ql-toolbar button:hover:not(.ql-active){color:#444}.ql-snow .ql-toolbar button:hover:not(.ql-active) .ql-fill,.ql-snow .ql-toolbar button:hover:not(.ql-active) .ql-stroke.ql-fill,.ql-snow.ql-toolbar button:hover:not(.ql-active) .ql-fill,.ql-snow.ql-toolbar button:hover:not(.ql-active) .ql-stroke.ql-fill{fill:#444}.ql-snow .ql-toolbar button:hover:not(.ql-active) .ql-stroke,.ql-snow .ql-toolbar button:hover:not(.ql-active) .ql-stroke-miter,.ql-snow.ql-toolbar button:hover:not(.ql-active) .ql-stroke,.ql-snow.ql-toolbar button:hover:not(.ql-active) .ql-stroke-miter{stroke:#444}}.ql-snow,.ql-snow *{box-sizing:border-box}.ql-snow .ql-hidden{display:none}.ql-snow .ql-out-bottom,.ql-snow .ql-out-top{visibility:hidden}.ql-snow .ql-tooltip{position:absolute;-webkit-transform:translateY(10px);transform:translateY(10px)}.ql-snow .ql-tooltip a{cursor:pointer;text-decoration:none;line-height:26px}.ql-snow .ql-tooltip.ql-flip{-webkit-transform:translateY(-10px);transform:translateY(-10px)}.ql-snow .ql-formats{display:inline-block;vertical-align:middle}.ql-snow .ql-formats:after{clear:both;content:'';display:table}.ql-snow .ql-stroke{fill:none;stroke:#444;stroke-linecap:round;stroke-linejoin:round;stroke-width:2}.ql-snow .ql-stroke-miter{fill:none;stroke:#444;stroke-miterlimit:10;stroke-width:2}.ql-snow .ql-fill,.ql-snow .ql-stroke.ql-fill{fill:#444}.ql-snow .ql-empty{fill:none}.ql-snow .ql-even{fill-rule:evenodd}.ql-snow .ql-stroke.ql-thin,.ql-snow .ql-thin{stroke-width:1}.ql-snow .ql-transparent{opacity:.4}.ql-snow .ql-direction svg:last-child{display:none}.ql-snow .ql-direction.ql-active svg:last-child{display:inline}.ql-snow .ql-direction.ql-active svg:first-child{display:none}.ql-snow .ql-editor h1{font-size:2em}.ql-snow .ql-editor h2{font-size:1.5em}.ql-snow .ql-editor h3{font-size:1.17em}.ql-snow .ql-editor h4{font-size:1em}.ql-snow .ql-editor h5{font-size:.83em}.ql-snow .ql-editor h6{font-size:.67em}.ql-snow .ql-editor a{text-decoration:underline}.ql-snow .ql-editor blockquote{border-left:4px solid #ccc;margin-bottom:5px;margin-top:5px;padding-left:16px}.ql-snow .ql-editor code,.ql-snow .ql-editor pre{background-color:#f0f0f0;border-radius:3px}.ql-snow .ql-editor pre{white-space:pre-wrap;margin-bottom:5px;margin-top:5px;padding:5px 10px}.ql-snow .ql-editor code{font-size:85%;padding:2px 4px}.ql-snow .ql-editor pre.ql-syntax{background-color:#23241f;color:#f8f8f2;overflow:visible}.ql-snow .ql-editor img{max-width:100%}.ql-snow .ql-picker{color:#444;display:inline-block;float:left;font-size:14px;font-weight:500;height:24px;position:relative;vertical-align:middle}.ql-snow .ql-picker-label{cursor:pointer;display:inline-block;height:100%;padding-left:8px;padding-right:2px;position:relative;width:100%;overflow:hidden}.ql-snow .ql-picker-options{background-color:#fff;display:none;min-width:100%;padding:4px 8px;position:absolute;white-space:nowrap}.ql-snow .ql-picker-options .ql-picker-item{cursor:pointer;display:block;padding-bottom:5px;padding-top:5px}.ql-snow .ql-picker.ql-expanded .ql-picker-label{color:#ccc;z-index:2}.ql-snow .ql-picker.ql-expanded .ql-picker-label .ql-fill{fill:#ccc}.ql-snow .ql-picker.ql-expanded .ql-picker-label .ql-stroke{stroke:#ccc}.ql-snow .ql-picker.ql-expanded .ql-picker-options{display:block;margin-top:-1px;top:100%;z-index:1}.ql-snow .ql-color-picker,.ql-snow .ql-icon-picker{width:28px}.ql-snow .ql-color-picker .ql-picker-label,.ql-snow .ql-icon-picker .ql-picker-label{padding:0 4px}.ql-snow .ql-color-picker .ql-picker-label svg,.ql-snow .ql-icon-picker .ql-picker-label svg{top:-1px;position:relative}.ql-snow .ql-icon-picker .ql-picker-options{padding:4px 0}.ql-snow .ql-icon-picker .ql-picker-item{height:24px;width:24px;padding:2px 4px}.ql-snow .ql-color-picker .ql-picker-options{padding:3px 5px;width:152px}.ql-snow .ql-color-picker .ql-picker-item{border:1px solid transparent;float:left;height:16px;margin:2px;padding:0;width:16px}.ql-snow .ql-picker:not(.ql-color-picker):not(.ql-icon-picker) svg{position:absolute;margin-top:-9px;right:0;top:50%;width:18px;background-color:#fff}.ql-snow .ql-picker.ql-font .ql-picker-item[data-label]:not([data-label=''])::before,.ql-snow .ql-picker.ql-font .ql-picker-label[data-label]:not([data-label=''])::before,.ql-snow .ql-picker.ql-header .ql-picker-item[data-label]:not([data-label=''])::before,.ql-snow .ql-picker.ql-header .ql-picker-label[data-label]:not([data-label=''])::before,.ql-snow .ql-picker.ql-size .ql-picker-item[data-label]:not([data-label=''])::before,.ql-snow .ql-picker.ql-size .ql-picker-label[data-label]:not([data-label=''])::before{content:attr(data-label)}.ql-snow .ql-picker.ql-header{width:98px}.ql-snow .ql-picker.ql-header .ql-picker-item::before,.ql-snow .ql-picker.ql-header .ql-picker-label::before{content:'Normal'}.ql-snow .ql-picker.ql-header .ql-picker-item[data-value='1']::before,.ql-snow .ql-picker.ql-header .ql-picker-label[data-value='1']::before{content:'Heading 1'}.ql-snow .ql-picker.ql-header .ql-picker-item[data-value='2']::before,.ql-snow .ql-picker.ql-header .ql-picker-label[data-value='2']::before{content:'Heading 2'}.ql-snow .ql-picker.ql-header .ql-picker-item[data-value='3']::before,.ql-snow .ql-picker.ql-header .ql-picker-label[data-value='3']::before{content:'Heading 3'}.ql-snow .ql-picker.ql-header .ql-picker-item[data-value='4']::before,.ql-snow .ql-picker.ql-header .ql-picker-label[data-value='4']::before{content:'Heading 4'}.ql-snow .ql-picker.ql-header .ql-picker-item[data-value='5']::before,.ql-snow .ql-picker.ql-header .ql-picker-label[data-value='5']::before{content:'Heading 5'}.ql-snow .ql-picker.ql-header .ql-picker-item[data-value='6']::before,.ql-snow .ql-picker.ql-header .ql-picker-label[data-value='6']::before{content:'Heading 6'}.ql-snow .ql-picker.ql-header .ql-picker-item[data-value='1']::before{font-size:2em}.ql-snow .ql-picker.ql-header .ql-picker-item[data-value='2']::before{font-size:1.5em}.ql-snow .ql-picker.ql-header .ql-picker-item[data-value='3']::before{font-size:1.17em}.ql-snow .ql-picker.ql-header .ql-picker-item[data-value='4']::before{font-size:1em}.ql-snow .ql-picker.ql-header .ql-picker-item[data-value='5']::before{font-size:.83em}.ql-snow .ql-picker.ql-header .ql-picker-item[data-value='6']::before{font-size:.67em}.ql-snow .ql-picker.ql-font{width:108px}.ql-snow .ql-picker.ql-font .ql-picker-item::before,.ql-snow .ql-picker.ql-font .ql-picker-label::before{content:'Sans Serif'}.ql-snow .ql-picker.ql-font .ql-picker-item[data-value=serif]::before,.ql-snow .ql-picker.ql-font .ql-picker-label[data-value=serif]::before{content:'Serif'}.ql-snow .ql-picker.ql-font .ql-picker-item[data-value=monospace]::before,.ql-snow .ql-picker.ql-font .ql-picker-label[data-value=monospace]::before{content:'Monospace'}.ql-snow .ql-picker.ql-font .ql-picker-item[data-value=serif]::before{font-family:Georgia,Times New Roman,serif}.ql-snow .ql-picker.ql-font .ql-picker-item[data-value=monospace]::before{font-family:Monaco,Courier New,monospace}.ql-snow .ql-picker.ql-size{width:98px}.ql-snow .ql-picker.ql-size .ql-picker-item::before,.ql-snow .ql-picker.ql-size .ql-picker-label::before{content:'Normal'}.ql-snow .ql-picker.ql-size .ql-picker-item[data-value=small]::before,.ql-snow .ql-picker.ql-size .ql-picker-label[data-value=small]::before{content:'Small'}.ql-snow .ql-picker.ql-size .ql-picker-item[data-value=large]::before,.ql-snow .ql-picker.ql-size .ql-picker-label[data-value=large]::before{content:'Large'}.ql-snow .ql-picker.ql-size .ql-picker-item[data-value=huge]::before,.ql-snow .ql-picker.ql-size .ql-picker-label[data-value=huge]::before{content:'Huge'}.ql-snow .ql-picker.ql-size .ql-picker-item[data-value=small]::before{font-size:10px}.ql-snow .ql-picker.ql-size .ql-picker-item[data-value=large]::before{font-size:18px}.ql-snow .ql-picker.ql-size .ql-picker-item[data-value=huge]::before{font-size:32px}.ql-snow .ql-color-picker.ql-background .ql-picker-item{background-color:#fff}.ql-snow .ql-color-picker.ql-color .ql-picker-item{background-color:#000}.ql-toolbar.ql-snow{border:1px solid #ccc;box-sizing:border-box;font-family:'Helvetica Neue',Helvetica,Arial,sans-serif;padding:8px}.ql-toolbar.ql-snow .ql-formats{margin-right:15px}.ql-toolbar.ql-snow .ql-picker-label{border:1px solid transparent}.ql-toolbar.ql-snow .ql-picker-options{border:1px solid transparent;box-shadow:rgba(0,0,0,.2) 0 2px 8px}.ql-toolbar.ql-snow .ql-picker.ql-expanded .ql-picker-label,.ql-toolbar.ql-snow .ql-picker.ql-expanded .ql-picker-options{border-color:#ccc}.ql-toolbar.ql-snow .ql-color-picker .ql-picker-item.ql-selected,.ql-toolbar.ql-snow .ql-color-picker .ql-picker-item:hover{border-color:#000}.ql-toolbar.ql-snow+.ql-container.ql-snow{border-top:0}.ql-snow .ql-tooltip{background-color:#fff;border:1px solid #ccc;box-shadow:0 0 5px #ddd;color:#444;padding:5px 12px;white-space:nowrap}.ql-snow .ql-tooltip::before{content:'Visit URL:';line-height:26px;margin-right:8px}.ql-snow .ql-tooltip input[type=text]{display:none;border:1px solid #ccc;font-size:13px;height:26px;margin:0;padding:3px 5px;width:170px}.ql-snow .ql-tooltip a.ql-preview{display:inline-block;max-width:200px;overflow-x:hidden;text-overflow:ellipsis;vertical-align:top}.ql-snow .ql-tooltip a.ql-action::after{border-right:1px solid #ccc;content:'Edit';margin-left:16px;padding-right:8px}.ql-snow .ql-tooltip a.ql-remove::before{content:'Remove';margin-left:8px}.ql-snow .ql-tooltip.ql-editing a.ql-preview,.ql-snow .ql-tooltip.ql-editing a.ql-remove{display:none}.ql-snow .ql-tooltip.ql-editing input[type=text]{display:inline-block}.ql-snow .ql-tooltip.ql-editing a.ql-action::after{border-right:0;content:'Save';padding-right:0}.ql-snow .ql-tooltip[data-mode=link]::before{content:'Enter link:'}.ql-snow .ql-tooltip[data-mode=formula]::before{content:'Enter formula:'}.ql-snow .ql-tooltip[data-mode=video]::before{content:'Enter video:'}.ql-snow a{color:#06c}.ql-container.ql-snow{border:1px solid #d9d9d9;border-top:1px solid #d9d9d9!important}.ql-toolbar.ql-snow{border-bottom:0;line-height:20px}.ql-snow .ql-picker.ql-font .ql-picker-item[data-value=SimSun]::before,.ql-snow .ql-picker.ql-font .ql-picker-label[data-value=SimSun]::before{content:'\u5B8B\u4F53';font-family:SimSun}.ql-snow .ql-picker.ql-font .ql-picker-item[data-value=SimHei]::before,.ql-snow .ql-picker.ql-font .ql-picker-label[data-value=SimHei]::before{content:'\u9ED1\u4F53';font-family:SimHei}.ql-snow .ql-picker.ql-font .ql-picker-item[data-value=Microsoft-YaHei]::before,.ql-snow .ql-picker.ql-font .ql-picker-label[data-value=Microsoft-YaHei]::before{content:'\u5FAE\u8F6F\u96C5\u9ED1';font-family:'Microsoft YaHei'}.ql-snow .ql-picker.ql-font .ql-picker-item[data-value=KaiTi]::before,.ql-snow .ql-picker.ql-font .ql-picker-label[data-value=KaiTi]::before{content:'\u6977\u4F53';font-family:KaiTi}.ql-snow .ql-picker.ql-font .ql-picker-item[data-value=FangSong]::before,.ql-snow .ql-picker.ql-font .ql-picker-label[data-value=FangSong]::before{content:'\u4EFF\u5B8B';font-family:FangSong}.ql-snow .ql-picker.ql-font .ql-picker-item[data-value=Arial]::before,.ql-snow .ql-picker.ql-font .ql-picker-label[data-value=Arial]::before{content:'Arial';font-family:Arial}.ql-snow .ql-picker.ql-font .ql-picker-item[data-value=Times]::before,.ql-snow .ql-picker.ql-font .ql-picker-label[data-value=Times]::before{content:'Times New Roman';font-family:'Times New Roman';white-space:nowrap}.ql-snow .ql-picker.ql-font .ql-picker-item[data-value=sans-serif]::before,.ql-snow .ql-picker.ql-font .ql-picker-label[data-value=sans-serif]::before{content:'sans-serif';font-family:sans-serif}.ql-font-SimSun{font-family:SimSun}.ql-font-SimHei{font-family:SimHei}.ql-font-Microsoft-YaHei{font-family:'Microsoft YaHei'}.ql-font-KaiTi{font-family:KaiTi}.ql-font-FangSong{font-family:FangSong}.ql-font-Arial{font-family:Arial}.ql-font-Times{font-family:'Times New Roman'}.ql-font-sans-serif{font-family:sans-serif}.ql-toolbar [tooltip]{position:relative}.ql-toolbar [tooltip]::before{content:'';position:absolute;top:-4px;left:50%;-webkit-transform:translateX(-50%);transform:translateX(-50%);border-width:4px 6px 0;border-style:solid;border-color:rgba(0,0,0,.7) transparent transparent;z-index:99;opacity:0}.ql-toolbar [tooltip-position=left]::before{left:0;top:50%;margin-left:-12px;-webkit-transform:translatey(-50%) rotate(-90deg);transform:translatey(-50%) rotate(-90deg)}.ql-toolbar [tooltip-position=top]::before{left:50%}.ql-toolbar [tooltip-position=bottom]::before{top:100%;margin-top:8px;-webkit-transform:translateX(-50%) translatey(-100%) rotate(-180deg);transform:translateX(-50%) translatey(-100%) rotate(-180deg)}.ql-toolbar [tooltip-position=right]::before{left:100%;top:50%;margin-left:1px;-webkit-transform:translatey(-50%) rotate(90deg);transform:translatey(-50%) rotate(90deg)}.ql-toolbar [tooltip]::after{content:attr(tooltip);position:absolute;left:50%;top:-4px;-webkit-transform:translateX(-50%) translateY(-100%);transform:translateX(-50%) translateY(-100%);background:rgba(0,0,0,.7);text-align:center;color:#fff;padding:4px;font-size:12px;min-width:70px;border-radius:5px;pointer-events:none;z-index:99;opacity:0}.ql-toolbar [tooltip-position=left]::after{left:0;top:50%;margin-left:-8px;-webkit-transform:translateX(-100%) translateY(-50%);transform:translateX(-100%) translateY(-50%)}.ql-toolbar [tooltip-position=top]::after{left:50%}.ql-toolbar [tooltip-position=bottom]::after{top:100%;margin-top:8px;-webkit-transform:translateX(-50%) translateY(0);transform:translateX(-50%) translateY(0)}.ql-toolbar [tooltip-position=right]::after{left:100%;top:50%;margin-left:8px;-webkit-transform:translateX(0) translateY(-50%);transform:translateX(0) translateY(-50%)}.ql-toolbar [tooltip]:hover::after,.ql-toolbar [tooltip]:hover::before{opacity:1}.quill-tooltip::before{content:'';position:absolute;bottom:-4px;left:50%;-webkit-transform:translateX(-50%);transform:translateX(-50%);border-width:4px 6px 0;border-style:solid;border-color:rgba(0,0,0,.7) transparent transparent;z-index:99}.ql-container.ql-snow,.ql-toolbar.ql-snow{border:none}.farris-html-editor{height:auto;border:1px solid #d9d9d9;display:block;padding:0}.farris-html-editor.farris-html-editor-disabled,.farris-html-editor.farris-html-editor-readonly{border:none}.ql-container.ql-disabled{border-top:none!important}.ql-bubble .ql-toolbar:after,.ql-bubble.ql-toolbar:after{clear:both;content:'';display:table}.ql-bubble .ql-toolbar button,.ql-bubble.ql-toolbar button{background:0 0;border:none;cursor:pointer;display:inline-block;float:left;height:24px;padding:3px 5px;width:28px}.ql-bubble .ql-toolbar button svg,.ql-bubble.ql-toolbar button svg{float:left;height:100%}.ql-bubble .ql-toolbar button:active:hover,.ql-bubble.ql-toolbar button:active:hover{outline:0}.ql-bubble .ql-toolbar input.ql-image[type=file],.ql-bubble.ql-toolbar input.ql-image[type=file]{display:none}.ql-bubble .ql-toolbar .ql-picker-item.ql-selected,.ql-bubble .ql-toolbar .ql-picker-item:hover,.ql-bubble .ql-toolbar .ql-picker-label.ql-active,.ql-bubble .ql-toolbar .ql-picker-label:hover,.ql-bubble .ql-toolbar button.ql-active,.ql-bubble .ql-toolbar button:focus,.ql-bubble .ql-toolbar button:hover,.ql-bubble.ql-toolbar .ql-picker-item.ql-selected,.ql-bubble.ql-toolbar .ql-picker-item:hover,.ql-bubble.ql-toolbar .ql-picker-label.ql-active,.ql-bubble.ql-toolbar .ql-picker-label:hover,.ql-bubble.ql-toolbar button.ql-active,.ql-bubble.ql-toolbar button:focus,.ql-bubble.ql-toolbar button:hover{color:#fff}.ql-bubble .ql-toolbar .ql-picker-item.ql-selected .ql-fill,.ql-bubble .ql-toolbar .ql-picker-item.ql-selected .ql-stroke.ql-fill,.ql-bubble .ql-toolbar .ql-picker-item:hover .ql-fill,.ql-bubble .ql-toolbar .ql-picker-item:hover .ql-stroke.ql-fill,.ql-bubble .ql-toolbar .ql-picker-label.ql-active .ql-fill,.ql-bubble .ql-toolbar .ql-picker-label.ql-active .ql-stroke.ql-fill,.ql-bubble .ql-toolbar .ql-picker-label:hover .ql-fill,.ql-bubble .ql-toolbar .ql-picker-label:hover .ql-stroke.ql-fill,.ql-bubble .ql-toolbar button.ql-active .ql-fill,.ql-bubble .ql-toolbar button.ql-active .ql-stroke.ql-fill,.ql-bubble .ql-toolbar button:focus .ql-fill,.ql-bubble .ql-toolbar button:focus .ql-stroke.ql-fill,.ql-bubble .ql-toolbar button:hover .ql-fill,.ql-bubble .ql-toolbar button:hover .ql-stroke.ql-fill,.ql-bubble.ql-toolbar .ql-picker-item.ql-selected .ql-fill,.ql-bubble.ql-toolbar .ql-picker-item.ql-selected .ql-stroke.ql-fill,.ql-bubble.ql-toolbar .ql-picker-item:hover .ql-fill,.ql-bubble.ql-toolbar .ql-picker-item:hover .ql-stroke.ql-fill,.ql-bubble.ql-toolbar .ql-picker-label.ql-active .ql-fill,.ql-bubble.ql-toolbar .ql-picker-label.ql-active .ql-stroke.ql-fill,.ql-bubble.ql-toolbar .ql-picker-label:hover .ql-fill,.ql-bubble.ql-toolbar .ql-picker-label:hover .ql-stroke.ql-fill,.ql-bubble.ql-toolbar button.ql-active .ql-fill,.ql-bubble.ql-toolbar button.ql-active .ql-stroke.ql-fill,.ql-bubble.ql-toolbar button:focus .ql-fill,.ql-bubble.ql-toolbar button:focus .ql-stroke.ql-fill,.ql-bubble.ql-toolbar button:hover .ql-fill,.ql-bubble.ql-toolbar button:hover .ql-stroke.ql-fill{fill:#fff}.ql-bubble .ql-toolbar .ql-picker-item.ql-selected .ql-stroke,.ql-bubble .ql-toolbar .ql-picker-item.ql-selected .ql-stroke-miter,.ql-bubble .ql-toolbar .ql-picker-item:hover .ql-stroke,.ql-bubble .ql-toolbar .ql-picker-item:hover .ql-stroke-miter,.ql-bubble .ql-toolbar .ql-picker-label.ql-active .ql-stroke,.ql-bubble .ql-toolbar .ql-picker-label.ql-active .ql-stroke-miter,.ql-bubble .ql-toolbar .ql-picker-label:hover .ql-stroke,.ql-bubble .ql-toolbar .ql-picker-label:hover .ql-stroke-miter,.ql-bubble .ql-toolbar button.ql-active .ql-stroke,.ql-bubble .ql-toolbar button.ql-active .ql-stroke-miter,.ql-bubble .ql-toolbar button:focus .ql-stroke,.ql-bubble .ql-toolbar button:focus .ql-stroke-miter,.ql-bubble .ql-toolbar button:hover .ql-stroke,.ql-bubble .ql-toolbar button:hover .ql-stroke-miter,.ql-bubble.ql-toolbar .ql-picker-item.ql-selected .ql-stroke,.ql-bubble.ql-toolbar .ql-picker-item.ql-selected .ql-stroke-miter,.ql-bubble.ql-toolbar .ql-picker-item:hover .ql-stroke,.ql-bubble.ql-toolbar .ql-picker-item:hover .ql-stroke-miter,.ql-bubble.ql-toolbar .ql-picker-label.ql-active .ql-stroke,.ql-bubble.ql-toolbar .ql-picker-label.ql-active .ql-stroke-miter,.ql-bubble.ql-toolbar .ql-picker-label:hover .ql-stroke,.ql-bubble.ql-toolbar .ql-picker-label:hover .ql-stroke-miter,.ql-bubble.ql-toolbar button.ql-active .ql-stroke,.ql-bubble.ql-toolbar button.ql-active .ql-stroke-miter,.ql-bubble.ql-toolbar button:focus .ql-stroke,.ql-bubble.ql-toolbar button:focus .ql-stroke-miter,.ql-bubble.ql-toolbar button:hover .ql-stroke,.ql-bubble.ql-toolbar button:hover .ql-stroke-miter{stroke:#fff}@media (pointer:coarse){.ql-bubble .ql-toolbar button:hover:not(.ql-active),.ql-bubble.ql-toolbar button:hover:not(.ql-active){color:#ccc}.ql-bubble .ql-toolbar button:hover:not(.ql-active) .ql-fill,.ql-bubble .ql-toolbar button:hover:not(.ql-active) .ql-stroke.ql-fill,.ql-bubble.ql-toolbar button:hover:not(.ql-active) .ql-fill,.ql-bubble.ql-toolbar button:hover:not(.ql-active) .ql-stroke.ql-fill{fill:#ccc}.ql-bubble .ql-toolbar button:hover:not(.ql-active) .ql-stroke,.ql-bubble .ql-toolbar button:hover:not(.ql-active) .ql-stroke-miter,.ql-bubble.ql-toolbar button:hover:not(.ql-active) .ql-stroke,.ql-bubble.ql-toolbar button:hover:not(.ql-active) .ql-stroke-miter{stroke:#ccc}}.ql-bubble,.ql-bubble *{box-sizing:border-box}.ql-bubble .ql-hidden{display:none}.ql-bubble .ql-out-bottom,.ql-bubble .ql-out-top{visibility:hidden}.ql-bubble .ql-tooltip{position:absolute;-webkit-transform:translateY(10px);transform:translateY(10px)}.ql-bubble .ql-tooltip a{cursor:pointer;text-decoration:none}.ql-bubble .ql-tooltip.ql-flip{-webkit-transform:translateY(-10px);transform:translateY(-10px)}.ql-bubble .ql-formats{display:inline-block;vertical-align:middle}.ql-bubble .ql-formats:after{clear:both;content:'';display:table}.ql-bubble .ql-stroke{fill:none;stroke:#ccc;stroke-linecap:round;stroke-linejoin:round;stroke-width:2}.ql-bubble .ql-stroke-miter{fill:none;stroke:#ccc;stroke-miterlimit:10;stroke-width:2}.ql-bubble .ql-fill,.ql-bubble .ql-stroke.ql-fill{fill:#ccc}.ql-bubble .ql-empty{fill:none}.ql-bubble .ql-even{fill-rule:evenodd}.ql-bubble .ql-stroke.ql-thin,.ql-bubble .ql-thin{stroke-width:1}.ql-bubble .ql-transparent{opacity:.4}.ql-bubble .ql-direction svg:last-child{display:none}.ql-bubble .ql-direction.ql-active svg:last-child{display:inline}.ql-bubble .ql-direction.ql-active svg:first-child{display:none}.ql-bubble .ql-editor h1{font-size:2em}.ql-bubble .ql-editor h2{font-size:1.5em}.ql-bubble .ql-editor h3{font-size:1.17em}.ql-bubble .ql-editor h4{font-size:1em}.ql-bubble .ql-editor h5{font-size:.83em}.ql-bubble .ql-editor h6{font-size:.67em}.ql-bubble .ql-editor a{text-decoration:underline}.ql-bubble .ql-editor blockquote{border-left:4px solid #ccc;margin-bottom:5px;margin-top:5px;padding-left:16px}.ql-bubble .ql-editor code,.ql-bubble .ql-editor pre{background-color:#f0f0f0;border-radius:3px}.ql-bubble .ql-editor pre{white-space:pre-wrap;margin-bottom:5px;margin-top:5px;padding:5px 10px}.ql-bubble .ql-editor code{font-size:85%;padding:2px 4px}.ql-bubble .ql-editor pre.ql-syntax{background-color:#23241f;color:#f8f8f2;overflow:visible}.ql-bubble .ql-editor img{max-width:100%}.ql-bubble .ql-picker{color:#ccc;display:inline-block;float:left;font-size:14px;font-weight:500;height:24px;position:relative;vertical-align:middle}.ql-bubble .ql-picker-label{cursor:pointer;display:inline-block;height:100%;padding-left:8px;padding-right:2px;position:relative;width:100%;overflow:hidden}.ql-bubble .ql-picker-label::before{display:inline-block;line-height:22px}.ql-bubble .ql-picker-options{background-color:#444;display:none;min-width:100%;padding:4px 8px;position:absolute;white-space:nowrap}.ql-bubble .ql-picker-options .ql-picker-item{cursor:pointer;display:block;padding-bottom:5px;padding-top:5px}.ql-bubble .ql-picker.ql-expanded .ql-picker-label{color:#777;z-index:2}.ql-bubble .ql-picker.ql-expanded .ql-picker-label .ql-fill{fill:#777}.ql-bubble .ql-picker.ql-expanded .ql-picker-label .ql-stroke{stroke:#777}.ql-bubble .ql-picker.ql-expanded .ql-picker-options{display:block;margin-top:-1px;top:100%;z-index:1}.ql-bubble .ql-color-picker,.ql-bubble .ql-icon-picker{width:28px}.ql-bubble .ql-color-picker .ql-picker-label,.ql-bubble .ql-icon-picker .ql-picker-label{padding:0 4px}.ql-bubble .ql-color-picker .ql-picker-label svg,.ql-bubble .ql-icon-picker .ql-picker-label svg{right:4px}.ql-bubble .ql-icon-picker .ql-picker-options{padding:4px 0}.ql-bubble .ql-icon-picker .ql-picker-item{height:24px;width:24px;padding:2px 4px}.ql-bubble .ql-color-picker .ql-picker-options{padding:3px 5px;width:152px}.ql-bubble .ql-color-picker .ql-picker-item{border:1px solid transparent;float:left;height:16px;margin:2px;padding:0;width:16px}.ql-bubble .ql-picker:not(.ql-color-picker):not(.ql-icon-picker) svg{position:absolute;margin-top:-9px;right:0;top:50%;width:18px;background-color:#fff}.ql-bubble .ql-picker.ql-font .ql-picker-item[data-label]:not([data-label=''])::before,.ql-bubble .ql-picker.ql-font .ql-picker-label[data-label]:not([data-label=''])::before,.ql-bubble .ql-picker.ql-header .ql-picker-item[data-label]:not([data-label=''])::before,.ql-bubble .ql-picker.ql-header .ql-picker-label[data-label]:not([data-label=''])::before,.ql-bubble .ql-picker.ql-size .ql-picker-item[data-label]:not([data-label=''])::before,.ql-bubble .ql-picker.ql-size .ql-picker-label[data-label]:not([data-label=''])::before{content:attr(data-label)}.ql-bubble .ql-picker.ql-header{width:98px}.ql-bubble .ql-picker.ql-header .ql-picker-item::before,.ql-bubble .ql-picker.ql-header .ql-picker-label::before{content:'Normal'}.ql-bubble .ql-picker.ql-header .ql-picker-item[data-value=\"1\"]::before,.ql-bubble .ql-picker.ql-header .ql-picker-label[data-value=\"1\"]::before{content:'Heading 1'}.ql-bubble .ql-picker.ql-header .ql-picker-item[data-value=\"2\"]::before,.ql-bubble .ql-picker.ql-header .ql-picker-label[data-value=\"2\"]::before{content:'Heading 2'}.ql-bubble .ql-picker.ql-header .ql-picker-item[data-value=\"3\"]::before,.ql-bubble .ql-picker.ql-header .ql-picker-label[data-value=\"3\"]::before{content:'Heading 3'}.ql-bubble .ql-picker.ql-header .ql-picker-item[data-value=\"4\"]::before,.ql-bubble .ql-picker.ql-header .ql-picker-label[data-value=\"4\"]::before{content:'Heading 4'}.ql-bubble .ql-picker.ql-header .ql-picker-item[data-value=\"5\"]::before,.ql-bubble .ql-picker.ql-header .ql-picker-label[data-value=\"5\"]::before{content:'Heading 5'}.ql-bubble .ql-picker.ql-header .ql-picker-item[data-value=\"6\"]::before,.ql-bubble .ql-picker.ql-header .ql-picker-label[data-value=\"6\"]::before{content:'Heading 6'}.ql-bubble .ql-picker.ql-header .ql-picker-item[data-value=\"1\"]::before{font-size:2em}.ql-bubble .ql-picker.ql-header .ql-picker-item[data-value=\"2\"]::before{font-size:1.5em}.ql-bubble .ql-picker.ql-header .ql-picker-item[data-value=\"3\"]::before{font-size:1.17em}.ql-bubble .ql-picker.ql-header .ql-picker-item[data-value=\"4\"]::before{font-size:1em}.ql-bubble .ql-picker.ql-header .ql-picker-item[data-value=\"5\"]::before{font-size:.83em}.ql-bubble .ql-picker.ql-header .ql-picker-item[data-value=\"6\"]::before{font-size:.67em}.ql-bubble .ql-picker.ql-font{width:108px}.ql-bubble .ql-picker.ql-font .ql-picker-item::before,.ql-bubble .ql-picker.ql-font .ql-picker-label::before{content:'Sans Serif'}.ql-bubble .ql-picker.ql-font .ql-picker-item[data-value=serif]::before,.ql-bubble .ql-picker.ql-font .ql-picker-label[data-value=serif]::before{content:'Serif'}.ql-bubble .ql-picker.ql-font .ql-picker-item[data-value=monospace]::before,.ql-bubble .ql-picker.ql-font .ql-picker-label[data-value=monospace]::before{content:'Monospace'}.ql-bubble .ql-picker.ql-font .ql-picker-item[data-value=serif]::before{font-family:Georgia,Times New Roman,serif}.ql-bubble .ql-picker.ql-font .ql-picker-item[data-value=monospace]::before{font-family:Monaco,Courier New,monospace}.ql-bubble .ql-picker.ql-size{width:98px}.ql-bubble .ql-picker.ql-size .ql-picker-item::before,.ql-bubble .ql-picker.ql-size .ql-picker-label::before{content:'Normal'}.ql-bubble .ql-picker.ql-size .ql-picker-item[data-value=small]::before,.ql-bubble .ql-picker.ql-size .ql-picker-label[data-value=small]::before{content:'Small'}.ql-bubble .ql-picker.ql-size .ql-picker-item[data-value=large]::before,.ql-bubble .ql-picker.ql-size .ql-picker-label[data-value=large]::before{content:'Large'}.ql-bubble .ql-picker.ql-size .ql-picker-item[data-value=huge]::before,.ql-bubble .ql-picker.ql-size .ql-picker-label[data-value=huge]::before{content:'Huge'}.ql-bubble .ql-picker.ql-size .ql-picker-item[data-value=small]::before{font-size:10px}.ql-bubble .ql-picker.ql-size .ql-picker-item[data-value=large]::before{font-size:18px}.ql-bubble .ql-picker.ql-size .ql-picker-item[data-value=huge]::before{font-size:32px}.ql-bubble .ql-color-picker.ql-background .ql-picker-item{background-color:#fff}.ql-bubble .ql-color-picker.ql-color .ql-picker-item{background-color:#000}.ql-bubble .ql-toolbar .ql-formats{margin:8px 12px 8px 0}.ql-bubble .ql-toolbar .ql-formats:first-child{margin-left:12px}.ql-bubble .ql-color-picker svg{margin:1px}.ql-bubble .ql-color-picker .ql-picker-item.ql-selected,.ql-bubble .ql-color-picker .ql-picker-item:hover{border-color:#fff}.ql-bubble .ql-tooltip{background-color:#444;border-radius:25px;color:#fff}.ql-bubble .ql-tooltip-arrow{border-left:6px solid transparent;border-right:6px solid transparent;content:\" \";display:block;left:50%;margin-left:-6px;position:absolute}.ql-bubble .ql-tooltip:not(.ql-flip) .ql-tooltip-arrow{border-bottom:6px solid #444;top:-6px}.ql-bubble .ql-tooltip.ql-flip .ql-tooltip-arrow{border-top:6px solid #444;bottom:-6px}.ql-bubble .ql-tooltip.ql-editing .ql-tooltip-editor{display:block}.ql-bubble .ql-tooltip.ql-editing .ql-formats{visibility:hidden}.ql-bubble .ql-tooltip-editor{display:none}.ql-bubble .ql-tooltip-editor input[type=text]{background:0 0;border:none;color:#fff;font-size:13px;height:100%;outline:0;padding:10px 20px;position:absolute;width:100%}.ql-bubble .ql-tooltip-editor a{top:10px;position:absolute;right:20px}.ql-bubble .ql-tooltip-editor a:before{color:#ccc;content:\"\\D7\";font-size:16px;font-weight:700}.ql-container.ql-bubble:not(.ql-disabled) a{position:relative;white-space:nowrap}.ql-container.ql-bubble:not(.ql-disabled) a::before{background-color:#444;border-radius:15px;top:-5px;font-size:12px;color:#fff;content:attr(href);font-weight:400;overflow:hidden;padding:5px 15px;text-decoration:none;z-index:1}.ql-container.ql-bubble:not(.ql-disabled) a::after{border-top:6px solid #444;border-left:6px solid transparent;border-right:6px solid transparent;top:0;content:\" \";height:0;width:0}.ql-container.ql-bubble:not(.ql-disabled) a::after,.ql-container.ql-bubble:not(.ql-disabled) a::before{left:0;margin-left:50%;position:absolute;-webkit-transform:translate(-50%,-100%);transform:translate(-50%,-100%);-webkit-transition:visibility .2s;transition:visibility .2s;visibility:hidden}.ql-container.ql-bubble:not(.ql-disabled) a:hover::after,.ql-container.ql-bubble:not(.ql-disabled) a:hover::before{visibility:visible}"]
            }] }
];
/** @nocollapse */
HtmlEditorComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: DomSanitizer },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
    { type: Renderer2 },
    { type: NgZone },
    { type: undefined, decorators: [{ type: Inject, args: [QUILL_CONFIG_TOKEN,] }] },
    { type: NotifyService, decorators: [{ type: Optional }] },
    { type: LocaleService, decorators: [{ type: Optional }] },
    { type: Injector, decorators: [{ type: Optional }] }
];
HtmlEditorComponent.propDecorators = {
    rootClass: [{ type: HostBinding, args: ['class',] }],
    readonly: [{ type: Input }, { type: HostBinding, args: ['class.farris-html-editor-readonly',] }],
    disabled: [{ type: Input }, { type: HostBinding, args: ['class.farris-html-editor-disabled',] }],
    showBorder: [{ type: Input }, { type: HostBinding, args: ['class.farris-html-editor-border',] }],
    format: [{ type: Input }],
    theme: [{ type: Input }],
    modules: [{ type: Input }],
    debug: [{ type: Input }],
    placeholder: [{ type: Input }],
    maxLength: [{ type: Input }],
    minLength: [{ type: Input }],
    required: [{ type: Input }],
    formats: [{ type: Input }],
    customToolbarPosition: [{ type: Input }],
    sanitize: [{ type: Input }],
    styles: [{ type: Input }],
    strict: [{ type: Input }],
    scrollingContainer: [{ type: Input }],
    bounds: [{ type: Input }],
    customOptions: [{ type: Input }],
    trackChanges: [{ type: Input }],
    preserveWhitespace: [{ type: Input }],
    limitImageSize: [{ type: Input }],
    maxImgSize: [{ type: Input }],
    customToolbar: [{ type: Input }],
    toolbar: [{ type: Input }],
    fontStyleBar: [{ type: Input }],
    BlockBar: [{ type: Input }],
    headerBar: [{ type: Input }],
    sizeBar: [{ type: Input }],
    listBar: [{ type: Input }],
    fontBar: [{ type: Input }],
    superAndSubscriptBar: [{ type: Input }],
    indentBar: [{ type: Input }],
    colorBar: [{ type: Input }],
    mediaBar: [{ type: Input }],
    imageBar: [{ type: Input }],
    videoBar: [{ type: Input }],
    cleanBar: [{ type: Input }],
    alignBar: [{ type: Input }],
    maxHeight: [{ type: Input }],
    minHeight: [{ type: Input }],
    autoHeight: [{ type: Input }],
    onEditorCreated: [{ type: Output }],
    onContentChanged: [{ type: Output }],
    onSelectionChanged: [{ type: Output }],
    onFocus: [{ type: Output }],
    onBlur: [{ type: Output }],
    imageUpload: [{ type: Input }],
    valueGetter: [{ type: Input }],
    valueSetter: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    HtmlEditorComponent.prototype.quillEditor;
    /** @type {?} */
    HtmlEditorComponent.prototype.editorElem;
    /** @type {?} */
    HtmlEditorComponent.prototype.content;
    /**
     * @type {?}
     * @private
     */
    HtmlEditorComponent.prototype._disabled;
    /** @type {?} */
    HtmlEditorComponent.prototype.rootClass;
    /** @type {?} */
    HtmlEditorComponent.prototype.readonly;
    /** @type {?} */
    HtmlEditorComponent.prototype.disabled;
    /** @type {?} */
    HtmlEditorComponent.prototype.showBorder;
    /** @type {?} */
    HtmlEditorComponent.prototype.format;
    /** @type {?} */
    HtmlEditorComponent.prototype.theme;
    /** @type {?} */
    HtmlEditorComponent.prototype.modules;
    /** @type {?} */
    HtmlEditorComponent.prototype.debug;
    /** @type {?} */
    HtmlEditorComponent.prototype.placeholder;
    /** @type {?} */
    HtmlEditorComponent.prototype.maxLength;
    /** @type {?} */
    HtmlEditorComponent.prototype.minLength;
    /** @type {?} */
    HtmlEditorComponent.prototype.required;
    /** @type {?} */
    HtmlEditorComponent.prototype.formats;
    /** @type {?} */
    HtmlEditorComponent.prototype.customToolbarPosition;
    /** @type {?} */
    HtmlEditorComponent.prototype.sanitize;
    /** @type {?} */
    HtmlEditorComponent.prototype.styles;
    /** @type {?} */
    HtmlEditorComponent.prototype.strict;
    /** @type {?} */
    HtmlEditorComponent.prototype.scrollingContainer;
    /** @type {?} */
    HtmlEditorComponent.prototype.bounds;
    /** @type {?} */
    HtmlEditorComponent.prototype.customOptions;
    /** @type {?} */
    HtmlEditorComponent.prototype.trackChanges;
    /** @type {?} */
    HtmlEditorComponent.prototype.preserveWhitespace;
    /**
     * 限制上传图片大小, 默认为 false不限制上传图片的大小
     * @type {?}
     */
    HtmlEditorComponent.prototype.limitImageSize;
    /**
     * 允许上传图片最大字节数，单位 kb
     * @type {?}
     */
    HtmlEditorComponent.prototype.maxImgSize;
    /** @type {?} */
    HtmlEditorComponent.prototype.customToolbar;
    /** @type {?} */
    HtmlEditorComponent.prototype.toolbar;
    /** @type {?} */
    HtmlEditorComponent.prototype.fontStyleBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.BlockBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.headerBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.sizeBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.listBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.fontBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.superAndSubscriptBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.indentBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.colorBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.mediaBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.imageBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.videoBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.cleanBar;
    /** @type {?} */
    HtmlEditorComponent.prototype.alignBar;
    /**
     * 只读禁用时，最大高度，为 0 时，不限制最大高度; 默认 800px
     * @type {?}
     */
    HtmlEditorComponent.prototype.maxHeight;
    /** @type {?} */
    HtmlEditorComponent.prototype.minHeight;
    /** @type {?} */
    HtmlEditorComponent.prototype.autoHeight;
    /** @type {?} */
    HtmlEditorComponent.prototype.onEditorCreated;
    /** @type {?} */
    HtmlEditorComponent.prototype.onContentChanged;
    /** @type {?} */
    HtmlEditorComponent.prototype.onSelectionChanged;
    /** @type {?} */
    HtmlEditorComponent.prototype.onFocus;
    /** @type {?} */
    HtmlEditorComponent.prototype.onBlur;
    /** @type {?} */
    HtmlEditorComponent.prototype.imageUpload;
    /** @type {?} */
    HtmlEditorComponent.prototype.valueGetter;
    /** @type {?} */
    HtmlEditorComponent.prototype.valueSetter;
    /** @type {?} */
    HtmlEditorComponent.prototype.selectionChangeHandler;
    /** @type {?} */
    HtmlEditorComponent.prototype.textChangeHandler;
    /**
     * @type {?}
     * @private
     */
    HtmlEditorComponent.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    HtmlEditorComponent.prototype.domSanitizer;
    /**
     * @type {?}
     * @private
     */
    HtmlEditorComponent.prototype.doc;
    /**
     * @type {?}
     * @private
     */
    HtmlEditorComponent.prototype.platformId;
    /**
     * @type {?}
     * @private
     */
    HtmlEditorComponent.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    HtmlEditorComponent.prototype.zone;
    /**
     * @type {?}
     * @private
     */
    HtmlEditorComponent.prototype.config;
    /**
     * @type {?}
     * @private
     */
    HtmlEditorComponent.prototype.notifySer;
    /**
     * @type {?}
     * @private
     */
    HtmlEditorComponent.prototype.localeSer;
    /**
     * @type {?}
     * @private
     */
    HtmlEditorComponent.prototype.injector;
}
export { ɵ0 };
