// @farris/ui-html-editor/esm2015/lib/html-editor.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { HtmlEditorComponent } from './html-editor.component';
import { defaultModules } from './html-editor-defaults';
import { QUILL_CONFIG_TOKEN } from './html-editor.interfaces';
import { NotifyModule } from '@farris/ui-notify';
import { LocaleModule } from '@farris/ui-locale';
const ɵ0 = { modules: defaultModules };
export class FarrisHtmlEditorModule {
    /**
     * @param {?=} config
     * @return {?}
     */
    static forRoot(config) {
        return {
            ngModule: FarrisHtmlEditorModule,
            providers: [
                {
                    provide: QUILL_CONFIG_TOKEN,
                    // tslint:disable-next-line:only-arrow-functions
                    useValue: config || { modules: defaultModules }
                }
            ]
        };
    }
}
FarrisHtmlEditorModule.decorators = [
    { type: NgModule, args: [{
                declarations: [HtmlEditorComponent],
                imports: [
                    NotifyModule.forRoot(),
                    LocaleModule.forRoot()
                ],
                exports: [HtmlEditorComponent],
                providers: [
                    {
                        provide: QUILL_CONFIG_TOKEN,
                        useValue: ɵ0
                    }
                ]
            },] }
];
export { ɵ0 };
