// @farris/ui-datagrid/esm2015/lib/components/footer/datagrid-footer-cell.styler.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Renderer2, Injector, Input } from '@angular/core';
import { DatagridComponent } from '../../datagrid.component';
export class FooterCellStylerDirective {
    /**
     * @param {?} el
     * @param {?} render
     * @param {?} inject
     * @param {?} dg
     */
    constructor(el, render, inject, dg) {
        this.el = el;
        this.render = render;
        this.inject = inject;
        this.dg = dg;
        this.isGroupFooter = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.isGroupFooter) {
            this.setGroupFooterCellStyle(this.column);
        }
        else {
            this.buildCustomCellStyle(this.column, this.rowData);
        }
    }
    /**
     * @private
     * @param {?} column
     * @param {?} rowData
     * @return {?}
     */
    buildCustomCellStyle(column, rowData) {
        /** @type {?} */
        const tr = this.el.nativeElement.parentNode;
        if (tr && !tr.id) {
            /** @type {?} */
            const rowdataid = this.dg.dfs.primaryId(rowData);
            /** @type {?} */
            let fixed = '';
            if (tr.closest('.f-datagrid-body-fixed-left')) {
                fixed = '-left';
            }
            else if (tr.closest('.f-datagrid-body-fixed-right')) {
                fixed = '-right';
            }
            /** @type {?} */
            const trID = 'footer-row' + fixed + '_' + (rowdataid || this.rowIndex);
            tr.id = trID;
        }
        if (!this.el.nativeElement.id) {
            this.el.nativeElement.id = 'footer-td_' + this.rowIndex + '_' + this.column.field.replace(/\./g, '_');
        }
        if (column && column.footer && column.footer.styler) {
            /** @type {?} */
            const _rd = rowData;
            /** @type {?} */
            const styler = column.footer.styler;
            /** @type {?} */
            const cs = styler(_rd[column.field], _rd);
            if (cs && Object.keys(cs).length) {
                this.dg.renderCustomStyle(cs, this.el.nativeElement);
            }
            else {
                this.dg.dgs.removeCssRule('#' + this.el.nativeElement.id);
            }
        }
    }
    /**
     * @private
     * @param {?} col
     * @return {?}
     */
    setGroupFooterCellStyle(col) {
        // 分组合计行自定义样式 20210909
        if (col.groupFooter && col.groupFooter.styler) {
            /** @type {?} */
            const val = this.dg.commonUtils.getValue(col.field, this.rowData);
            /** @type {?} */
            const _customStyles = this.dg.dgs.getCustomStyle(col.groupFooter.styler, [val, this.rowData]);
            if (_customStyles) {
                this.dg.renderCustomCls(_customStyles.cls, this.el.nativeElement);
                if (_customStyles.style) {
                    Object.keys(_customStyles.style).forEach((/**
                     * @param {?} n
                     * @return {?}
                     */
                    n => {
                        this.el.nativeElement.style[n] = _customStyles.style[n];
                    }));
                }
            }
        }
    }
}
FooterCellStylerDirective.decorators = [
    { type: Directive, args: [{
                selector: '[footer-cell-styler]',
            },] }
];
/** @nocollapse */
FooterCellStylerDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: Injector },
    { type: DatagridComponent }
];
FooterCellStylerDirective.propDecorators = {
    column: [{ type: Input }],
    rowData: [{ type: Input }],
    rowIndex: [{ type: Input }],
    isGroupFooter: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FooterCellStylerDirective.prototype.column;
    /** @type {?} */
    FooterCellStylerDirective.prototype.rowData;
    /** @type {?} */
    FooterCellStylerDirective.prototype.rowIndex;
    /** @type {?} */
    FooterCellStylerDirective.prototype.isGroupFooter;
    /**
     * @type {?}
     * @private
     */
    FooterCellStylerDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FooterCellStylerDirective.prototype.render;
    /**
     * @type {?}
     * @private
     */
    FooterCellStylerDirective.prototype.inject;
    /** @type {?} */
    FooterCellStylerDirective.prototype.dg;
}
