// @farris/ui-forms/esm2015/lib/comments/single-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, HostListener, Injector, Input, Output, ViewChild } from '@angular/core';
export class SingleListComponent {
    /**
     * @param {?} injector
     * @param {?} cdr
     */
    constructor(injector, cdr) {
        this.injector = injector;
        this.cdr = cdr;
        this.showButtons = true;
        this.buttons = [];
        this.data = [];
        this.textField = '';
        this.maxItems = 10;
        this.emptyDataMsg = '暂无数据';
        this.itemClick = new EventEmitter();
        this.activeIndex = -1;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClick($event) {
        $event.stopPropagation();
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @param {?} items
     * @return {?}
     */
    loadData(items) {
        this.data = items;
        if (!this.cdr['destroyed']) {
            this.cdr.detectChanges();
        }
    }
    /**
     * @param {?} rowObj
     * @return {?}
     */
    formatData(rowObj) {
        return this.formatter(rowObj);
    }
    /**
     * @param {?} $event
     * @param {?} item
     * @return {?}
     */
    onItemClick($event, item) {
        this.itemClick.emit({ data: item, evnet: $event });
    }
    /**
     * @param {?} index
     * @return {?}
     */
    setActiveItem(index) {
        this.activeIndex = index;
        if (!this.cdr['destroyed']) {
            this.cdr.detectChanges();
        }
    }
}
SingleListComponent.decorators = [
    { type: Component, args: [{
                selector: 'single-list',
                template: "<div class=\"d-flex flex-column\" style=\"width: 100%;height:100%;padding-top: 5px; \" #itemsContainer>\r\n    <!-- <div class=\"header\"></div> -->\r\n    <div class=\"list f-utils-fill f-datalist\" style=\"overflow: auto;flex:1;  max-height: 250px;\">\r\n        <ul class=\"list-group list-group-flush p-0\"  [class.h-100]=\"!data || !data.length\">\r\n            <li class=\"list-group-item list-group-item-action\" style=\"cursor: pointer;\" [class.active]=\"activeIndex === i\" *ngFor=\"let item of data; index as i\" (click)=\"onItemClick($event, item)\">\r\n                <span style=\"overflow: hidden;text-overflow: ellipsis;\" title=\" {{item[textField]}}\" *ngIf=\"!formatter\">{{ item[textField] }}</span> \r\n                <span style=\"overflow: hidden;text-overflow: ellipsis;\" *ngIf=\"formatter\" [innerHTML]=\"formatData(item) | safe:'html'\"></span>\r\n            </li>\r\n\r\n            <li class=\"norecord p-2\" *ngIf=\"!data || !data.length\" style=\"color: #848c9a;\">\r\n                {{ emptyDataMsg }}\r\n            </li>\r\n        </ul>\r\n    </div>\r\n    <div class=\"more d-flex\" *ngIf=\"showButtons\" style=\"height: 36px;border-top: 1px solid #dedede;justify-content: center;align-items: center;\">\r\n        <button class=\"btn btn-link\" *ngFor=\"let btn of buttons\" (click)=\"btn.handler($event)\"> \r\n            <span style=\"font-size: 13px;\" class=\"{{btn.iconCls}}\" *ngIf=\"btn.iconCls\"></span>\r\n            {{ btn.text }}\r\n        </button>\r\n    </div>\r\n     <!-- <div class=\"footer\"></div> -->\r\n</div>"
            }] }
];
/** @nocollapse */
SingleListComponent.ctorParameters = () => [
    { type: Injector },
    { type: ChangeDetectorRef }
];
SingleListComponent.propDecorators = {
    showButtons: [{ type: Input }],
    buttons: [{ type: Input }],
    data: [{ type: Input }],
    textField: [{ type: Input }],
    formatter: [{ type: Input }],
    maxItems: [{ type: Input }],
    emptyDataMsg: [{ type: Input }],
    itemClick: [{ type: Output }],
    itemsContainer: [{ type: ViewChild, args: ['itemsContainer',] }],
    onClick: [{ type: HostListener, args: ['mousedown', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    SingleListComponent.prototype.showButtons;
    /** @type {?} */
    SingleListComponent.prototype.buttons;
    /** @type {?} */
    SingleListComponent.prototype.data;
    /** @type {?} */
    SingleListComponent.prototype.textField;
    /** @type {?} */
    SingleListComponent.prototype.formatter;
    /** @type {?} */
    SingleListComponent.prototype.maxItems;
    /** @type {?} */
    SingleListComponent.prototype.emptyDataMsg;
    /** @type {?} */
    SingleListComponent.prototype.itemClick;
    /** @type {?} */
    SingleListComponent.prototype.itemsContainer;
    /** @type {?} */
    SingleListComponent.prototype.activeIndex;
    /**
     * @type {?}
     * @private
     */
    SingleListComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    SingleListComponent.prototype.cdr;
}
