// @farris/mobile-devkit/esm5/lib/http/http-util.js
var HttpUtil = /** @class */ (function () {
    function HttpUtil() {
    }
    /**
     * 追加Header
     */
    HttpUtil.appendHeader = function (headers, key, value) {
        var _a;
        headers = Object.assign({}, headers, (_a = {}, _a[key] = value, _a));
        return headers;
    };
    /**
     * 向RequestConfig中追加body
     */
    HttpUtil.appendBodyToRequestConfig = function (body, requestConfig) {
        if (!requestConfig) {
            requestConfig = {};
        }
        requestConfig = Object.assign({}, requestConfig, { body: body });
        return requestConfig;
    };
    /**
     * 构造AxiosReqeustConfig
     */
    HttpUtil.buildAxiosRequestConfig = function (method, url, requestConfig) {
        requestConfig = requestConfig || {};
        var axiosRequestConfig = {
            url: url,
            method: method,
            params: requestConfig.params || null,
            headers: requestConfig.headers || null,
            responseType: requestConfig.responseType || 'json',
            data: requestConfig.body || null
        };
        return axiosRequestConfig;
    };
    /**
     * 构造Http响应信息
     */
    HttpUtil.buildHttpResponse = function (axiosResponse) {
        var httpResponse = {
            body: axiosResponse.data,
            headers: axiosResponse.headers,
            status: axiosResponse.status,
            statusText: axiosResponse.statusText
        };
        return httpResponse;
    };
    return HttpUtil;
}());
export { HttpUtil };
