// @gsp-lcm/metadatart-selector4biztype/esm2015/metadata/metadatart-select4biztype/components/metadatainfo/metadatainfo.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadatart-select4biztype/components/metadatainfo/metadatainfo.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// tslint:disable-next-line:max-line-length
import { Component, Output, EventEmitter } from '@angular/core';
import { MetadataInfoService } from '../../metadatainfo.service';
import { NotifyService } from '@farris/ui-notify';
import { TranslateService } from '@ngx-translate/core';
export class MetadataInfoComponent {
    /**
     * @param {?} metadataInfoService
     * @param {?} notifyService
     * @param {?} translate
     */
    constructor(metadataInfoService, notifyService, translate) {
        this.metadataInfoService = metadataInfoService;
        this.notifyService = notifyService;
        this.translate = translate;
        this.OK = this.transform('OK');
        this.cancel = this.transform('Cancel');
        this.selectedMetadata = new EventEmitter();
        this.closeModalEmitter = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} key
     * @return {?}
     */
    transform(key) {
        return this.translate.instant(key);
    }
    /**
     * @return {?}
     */
    LoadMetadata() {
        if (this.metadataInfoService.metadatas == null || this.metadataInfoService.metadatas.length <= 0) {
            this.notifyService.warning(this.transform('PleaseChooseData'));
            return;
        }
        /** @type {?} */
        const Id = this.metadataInfoService.metadatas;
        this.selectedMetadata.emit(this.metadataInfoService.metadatas);
        this.hideModal();
        // 清空缓存
        this.metadataInfoService.clearMetadataRT();
    }
    /**
     * @return {?}
     */
    hideModal() {
        this.closeModalEmitter.emit();
    }
    /**
     * @param {?} bo
     * @return {?}
     */
    onSelectedChange(bo) {
        if (bo.id === 'gscloud') {
            this.boId = '';
        }
        else {
            this.boId = bo.id;
        }
    }
}
MetadataInfoComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'app-metadatainfo',
                template: "<!-- <div class=\"modal-body\">\r\n    <div class=\"templateBlock\">\r\n        <app-treeclassifier></app-treeclassifier>\r\n        <app-metadatainfolist></app-metadatainfolist>\r\n    </div>\r\n</div> -->\r\n<div class=\"farris-main-area flex-column farris-overflow-hidden\">\r\n  <div class=\"f-utils-fill bg-c-white position-relative f-utils-overflow-xhya d-flex\">\r\n      <div class=\"templateBlock flex-column flex-fill d-flex w-100\">\r\n        <kendo-splitter class=\"d-flex h-100 \" style=\"flex: 1 1 auto;\" orientation=\"horizontal\">\r\n          <kendo-splitter-pane class=\"d-flex \" [collapsible]=\"true\" size=\"250px\" style=\"height: auto\">\r\n            <app-bo-tree-panel (selectionChanged)=onSelectedChange($event)></app-bo-tree-panel>\r\n          </kendo-splitter-pane>\r\n          <kendo-splitter-pane class=\"d-flex\" style=\"height: auto\">\r\n            <app-metadatainfolist class=\"flex-fill\" [boId]=\"boId\"></app-metadatainfolist>\r\n          </kendo-splitter-pane>\r\n        </kendo-splitter>\r\n      </div>\r\n    </div>\r\n    <div class=\"modal-footer\">\r\n      <button type=\"button\" class=\"btn btn-default btn-primary\" (click)=\"LoadMetadata()\">{{ 'OK' | translate }}</button>\r\n      <button type=\"button\" class=\"btn btn-default\" (click)=\"hideModal()\">{{ 'Cancel' | translate }}</button>\r\n  </div>\r\n</div>",
                styles: [".templateBlock{display:flex}.infoBlock{width:100%;min-width:400px}.infoBlock label{width:50px}"]
            }] }
];
/** @nocollapse */
MetadataInfoComponent.ctorParameters = () => [
    { type: MetadataInfoService },
    { type: NotifyService },
    { type: TranslateService }
];
MetadataInfoComponent.propDecorators = {
    selectedMetadata: [{ type: Output }],
    closeModalEmitter: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    MetadataInfoComponent.prototype.boId;
    /** @type {?} */
    MetadataInfoComponent.prototype.OK;
    /** @type {?} */
    MetadataInfoComponent.prototype.cancel;
    /** @type {?} */
    MetadataInfoComponent.prototype.selectedMetadata;
    /** @type {?} */
    MetadataInfoComponent.prototype.closeModalEmitter;
    /**
     * @type {?}
     * @private
     */
    MetadataInfoComponent.prototype.metadataInfoService;
    /**
     * @type {?}
     * @private
     */
    MetadataInfoComponent.prototype.notifyService;
    /**
     * @type {?}
     * @private
     */
    MetadataInfoComponent.prototype.translate;
}
