// @farris/ui-datepicker/esm2015/lib/constants/constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Constants
 * @type {?}
 */
const D = 'd';
/** @type {?} */
const DD = 'dd';
/** @type {?} */
const M = 'M';
/** @type {?} */
const MM = 'MM';
/** @type {?} */
const MMM = 'MMM';
/** @type {?} */
const Y = 'y';
/** @type {?} */
const YYYY = 'yyyy';
/** @type {?} */
const DATE_ROW_COUNT = 5;
/** @type {?} */
const DATE_COL_COUNT = 6;
/** @type {?} */
const MONTH_ROW_COUNT = 3;
/** @type {?} */
const MONTH_COL_COUNT = 2;
/** @type {?} */
const YEAR_ROW_COUNT = 4;
/** @type {?} */
const YEAR_COL_COUNT = 4;
/** @type {?} */
const DOT = '.';
/** @type {?} */
const UNDER_LINE = '_';
/** @type {?} */
const SU = 'su';
/** @type {?} */
const MO = 'mo';
/** @type {?} */
const TU = 'tu';
/** @type {?} */
const WE = 'we';
/** @type {?} */
const TH = 'th';
/** @type {?} */
const FR = 'fr';
/** @type {?} */
const SA = 'sa';
/** @type {?} */
const DEFAULT_LOCALE = 'ZH_CN';
/** @type {?} */
const ZERO_STR = '0';
/** @type {?} */
const EMPTY_STR = '';
/** @type {?} */
const CLICK = 'click';
/** @type {?} */
const KEYUP = 'keyup';
/** @type {?} */
const BLUR = 'blur';
/** @type {?} */
const DISABLED = 'disabled';
/** @type {?} */
const BODY = 'body';
/** @type {?} */
const VALUE = 'value';
/** @type {?} */
const OPTIONS = 'options';
/** @type {?} */
const DEFAULT_MONTH = 'defaultMonth';
/** @type {?} */
const LOCALE = 'locale';
/** @type {?} */
const OBJECT = 'object';
/** @type {?} */
const PX = 'px';
/** @type {?} */
const STYLE = 'style';
/** @type {?} */
const OPTS = 'opts';
/** @type {?} */
const YEARS = 'years';
/** @type {?} */
const VISIBLE_MONTH = 'visibleMonth';
/** @type {?} */
const SELECT_MONTH = 'selectMonth';
/** @type {?} */
const SELECT_YEAR = 'selectYear';
/** @type {?} */
const PREV_VIEW_DISABLED = 'prevViewDisabled';
/** @type {?} */
const NEXT_VIEW_DISABLED = 'nextViewDisabled';
/** @type {?} */
const DATES = 'dates';
/** @type {?} */
const WEEK_DAYS = 'weekDays';
/** @type {?} */
const SELECTED_DATE = 'selectedDate';
/** @type {?} */
const SELECTED_DATE_RANGE = 'selectedDateRange';
/** @type {?} */
const MONTHS = 'months';
/** @type {?} */
const PREVENT_CLOSE_TIMEOUT = 50;
export { D, DD, M, MM, MMM, Y, YYYY, DATE_ROW_COUNT, DATE_COL_COUNT, MONTH_ROW_COUNT, MONTH_COL_COUNT, YEAR_ROW_COUNT, YEAR_COL_COUNT, DOT, UNDER_LINE, SU, MO, TU, WE, TH, FR, SA, DEFAULT_LOCALE, ZERO_STR, EMPTY_STR, CLICK, KEYUP, BLUR, DISABLED, BODY, VALUE, OPTIONS, DEFAULT_MONTH, LOCALE, OBJECT, PX, STYLE, OPTS, YEARS, VISIBLE_MONTH, SELECT_MONTH, SELECT_YEAR, PREV_VIEW_DISABLED, NEXT_VIEW_DISABLED, DATES, WEEK_DAYS, SELECTED_DATE, SELECTED_DATE_RANGE, MONTHS, PREVENT_CLOSE_TIMEOUT };
