// @edp-bif/common-api/esm5/lib/entity/impl/AbstractBillWritebackProcessEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/AbstractBillWritebackProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '../../ref';
import { AbstractBillProcessEntity } from './AbstractBillProcessEntity';
/**
 * 单据回写流实体类。
 */
var /**
 * 单据回写流实体类。
 */
AbstractBillWritebackProcessEntity = /** @class */ (function (_super) {
    tslib_1.__extends(AbstractBillWritebackProcessEntity, _super);
    function AbstractBillWritebackProcessEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    AbstractBillWritebackProcessEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.writebackRuleId = null;
        this.writebackProcessId = null;
        this.upBillVoProxyMetadataId = null;
        this.upBillVoProxyId = null;
        this.upBillVoProxyMetadataRootId = null;
        this.upBillVoProxyMetadataParentId = null;
        this.upBillVoProxyMetadataIdPath = null;
        this.upBillVoProxyRootId = null;
        this.upBillVoProxyParentId = null;
        this.upBillVoProxyIdPath = null;
    };
    /**
     * @return {?}
     */
    AbstractBillWritebackProcessEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "writebackRuleId", this.writebackRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "writebackProcessId", this.writebackProcessId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyMetadataId", this.upBillVoProxyMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyId", this.upBillVoProxyId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyMetadataRootId", this.upBillVoProxyMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyMetadataParentId", this.upBillVoProxyMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyMetadataIdPath", this.upBillVoProxyMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyRootId", this.upBillVoProxyRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyParentId", this.upBillVoProxyParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyIdPath", this.upBillVoProxyIdPath);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AbstractBillWritebackProcessEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.writebackRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackRuleId");
        this.writebackProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackProcessId");
        this.upBillVoProxyMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyMetadataId");
        this.upBillVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyId");
        this.upBillVoProxyMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyMetadataRootId");
        this.upBillVoProxyMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyMetadataParentId");
        this.upBillVoProxyMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyMetadataIdPath");
        this.upBillVoProxyRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyRootId");
        this.upBillVoProxyParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyParentId");
        this.upBillVoProxyIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyIdPath");
    };
    return AbstractBillWritebackProcessEntity;
}(AbstractBillProcessEntity));
/**
 * 单据回写流实体类。
 */
export { AbstractBillWritebackProcessEntity };
if (false) {
    /** @type {?} */
    AbstractBillWritebackProcessEntity.prototype.upBillVoProxyMetadataRootId;
    /** @type {?} */
    AbstractBillWritebackProcessEntity.prototype.upBillVoProxyMetadataParentId;
    /** @type {?} */
    AbstractBillWritebackProcessEntity.prototype.upBillVoProxyMetadataIdPath;
    /** @type {?} */
    AbstractBillWritebackProcessEntity.prototype.upBillVoProxyRootId;
    /** @type {?} */
    AbstractBillWritebackProcessEntity.prototype.upBillVoProxyParentId;
    /** @type {?} */
    AbstractBillWritebackProcessEntity.prototype.upBillVoProxyIdPath;
    /** @type {?} */
    AbstractBillWritebackProcessEntity.prototype.upBillVoProxyMetadataId;
    /** @type {?} */
    AbstractBillWritebackProcessEntity.prototype.upBillVoProxyId;
    /** @type {?} */
    AbstractBillWritebackProcessEntity.prototype.writebackRuleId;
    /** @type {?} */
    AbstractBillWritebackProcessEntity.prototype.writebackProcessId;
}
