// @edp-bif/common-api/esm5/lib/entity/impl/AbstractBillProcessEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/AbstractBillProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Entity, BpmnModelJsonHelper } from '../../ref';
/**
 * 单据流实体类。
 */
var /**
 * 单据流实体类。
 */
AbstractBillProcessEntity = /** @class */ (function (_super) {
    tslib_1.__extends(AbstractBillProcessEntity, _super);
    function AbstractBillProcessEntity() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.enabledParentEventRouterCond = true;
        _this.enabledParentMappingCond = true;
        _this.sysInit = false;
        _this.origin = true;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    AbstractBillProcessEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.code = null;
        this.name = null;
        this.description = null;
        this.persistenceVersion = null;
        this.version = 0;
        this.callType = null;
        // this.bizTypeL1 = null;
        // this.bizTypeL2 = null;
        // this.bizTypeL3 = null;
        // this.bizTypeL4 = null;
        this.upBillAppCode = null;
        this.upBillSuCode = null;
        this.upBillBoId = null;
        this.upBillBeMetadataId = null;
        this.upBillBeId = null;
        this.upBillVoMetadataId = null;
        this.upBillVoId = null;
        this.downBillAppCode = null;
        this.downBillSuCode = null;
        this.downBillBoId = null;
        this.downBillBeMetadataId = null;
        this.downBillBeId = null;
        this.downBillVoMetadataId = null;
        this.downBillVoId = null;
        this.isSupportAutomatinFlow = false;
        this.bizEventId = null;
        this.bizActionId = null;
        this.originId = null;
        this.rootId = null;
        this.parentId = null;
        this.enabledParentEventRouterCond = true;
        this.enabledParentMappingCond = true;
        this.idPath = null;
        this.versionTag = null;
        this.upBillBeMetadataRootId = null;
        this.upBillBeMetadataParentId = null;
        this.upBillBeMetadataIdPath = null;
        this.upBillBeRootId = null;
        this.upBillBeParentId = null;
        this.upBillBeIdPath = null;
        this.upBillVoMetadataRootId = null;
        this.upBillVoMetadataParentId = null;
        this.upBillVoMetadataIdPath = null;
        this.upBillVoRootId = null;
        this.upBillVoParentId = null;
        this.upBillVoIdPath = null;
        this.downBillBeMetadataRootId = null;
        this.downBillBeMetadataParentId = null;
        this.downBillBeMetadataIdPath = null;
        this.downBillBeRootId = null;
        this.downBillBeParentId = null;
        this.downBillBeIdPath = null;
        this.downBillVoMetadataRootId = null;
        this.downBillVoMetadataParentId = null;
        this.downBillVoMetadataIdPath = null;
        this.downBillVoRootId = null;
        this.downBillVoParentId = null;
        this.downBillVoIdPath = null;
        this.sysInit = false;
        this.origin = true;
        this.colI18nValues = null;
    };
    /**
     * @return {?}
     */
    AbstractBillProcessEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "description", this.description);
        BpmnModelJsonHelper.SetAttributeValue(obj, "persistenceVersion", this.persistenceVersion);
        BpmnModelJsonHelper.SetAttributeValue(obj, "version", this.version);
        BpmnModelJsonHelper.SetAttributeValue(obj, "callType", this.callType);
        // BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeL1", this.bizTypeL1);
        // BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeL2", this.bizTypeL2);
        // BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeL3", this.bizTypeL3);
        // BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeL4", this.bizTypeL4);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillAppCode", this.upBillAppCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillSuCode", this.upBillSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBoId", this.upBillBoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeMetadataId", this.upBillBeMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeId", this.upBillBeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataId", this.upBillVoMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoId", this.upBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillAppCode", this.downBillAppCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillSuCode", this.downBillSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBoId", this.downBillBoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeMetadataId", this.downBillBeMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeId", this.downBillBeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataId", this.downBillVoMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoId", this.downBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportAutomatinFlow", this.isSupportAutomatinFlow);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventId", this.bizEventId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionId", this.bizActionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originId", this.originId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "rootId", this.rootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "parentId", this.parentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "enabledParentEventRouterCond", this.enabledParentEventRouterCond);
        BpmnModelJsonHelper.SetAttributeValue(obj, "enabledParentMappingCond", this.enabledParentMappingCond);
        BpmnModelJsonHelper.SetAttributeValue(obj, "idPath", this.idPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "versionTag", this.versionTag);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeMetadataRootId", this.upBillBeMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeMetadataParentId", this.upBillBeMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeMetadataIdPath", this.upBillBeMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeRootId", this.upBillBeRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeParentId", this.upBillBeParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeIdPath", this.upBillBeIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataRootId", this.upBillVoMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataParentId", this.upBillVoMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataIdPath", this.upBillVoMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoRootId", this.upBillVoRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoParentId", this.upBillVoParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoIdPath", this.upBillVoIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeMetadataRootId", this.downBillBeMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeMetadataParentId", this.downBillBeMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeMetadataIdPath", this.downBillBeMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeRootId", this.downBillBeRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeParentId", this.downBillBeParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeIdPath", this.downBillBeIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataRootId", this.downBillVoMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataParentId", this.downBillVoMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataIdPath", this.downBillVoMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoRootId", this.downBillVoRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoParentId", this.downBillVoParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoIdPath", this.downBillVoIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sysInit", this.sysInit);
        BpmnModelJsonHelper.SetAttributeValue(obj, "origin", this.origin);
        BpmnModelJsonHelper.SetAttributeValue(obj, "colI18nValues", this.colI18nValues);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AbstractBillProcessEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.description = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "description");
        this.persistenceVersion = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "persistenceVersion");
        this.version = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "version");
        this.callType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "callType");
        // this.bizTypeL1 = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeL1");
        // this.bizTypeL2 = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeL2");
        // this.bizTypeL3 = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeL3");
        // this.bizTypeL4 = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeL4");
        this.upBillAppCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillAppCode");
        this.upBillSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillSuCode");
        this.upBillBoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBoId");
        this.upBillBeMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeMetadataId");
        this.upBillBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeId");
        this.upBillVoMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataId");
        this.upBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoId");
        this.downBillAppCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillAppCode");
        this.downBillSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillSuCode");
        this.downBillBoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBoId");
        this.downBillBeMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeMetadataId");
        this.downBillBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeId");
        this.downBillVoMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataId");
        this.downBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoId");
        this.isSupportAutomatinFlow = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportAutomatinFlow");
        this.bizEventId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventId");
        this.bizActionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionId");
        this.originId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originId");
        this.rootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "rootId");
        this.parentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "parentId");
        this.enabledParentEventRouterCond = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "enabledParentEventRouterCond");
        this.enabledParentMappingCond = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "enabledParentMappingCond");
        this.idPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "idPath");
        this.versionTag = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "versionTag");
        this.upBillBeMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeMetadataRootId");
        this.upBillBeMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeMetadataParentId");
        this.upBillBeMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeMetadataIdPath");
        this.upBillBeRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeRootId");
        this.upBillBeParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeParentId");
        this.upBillBeIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeIdPath");
        this.upBillVoMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataRootId");
        this.upBillVoMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataParentId");
        this.upBillVoMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataIdPath");
        this.upBillVoRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoRootId");
        this.upBillVoParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoParentId");
        this.upBillVoIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoIdPath");
        this.downBillBeMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeMetadataRootId");
        this.downBillBeMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeMetadataParentId");
        this.downBillBeMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeMetadataIdPath");
        this.downBillBeRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeRootId");
        this.downBillBeParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeParentId");
        this.downBillBeIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeIdPath");
        this.downBillVoMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataRootId");
        this.downBillVoMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataParentId");
        this.downBillVoMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataIdPath");
        this.downBillVoRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoRootId");
        this.downBillVoParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoParentId");
        this.downBillVoIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoIdPath");
        this.sysInit = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "sysInit");
        this.origin = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "origin");
        this.colI18nValues = BpmnModelJsonHelper.GetAttributeValue(obj, "colI18nValues");
    };
    return AbstractBillProcessEntity;
}(Entity));
/**
 * 单据流实体类。
 */
export { AbstractBillProcessEntity };
if (false) {
    /** @type {?} */
    AbstractBillProcessEntity.prototype.code;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.name;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.description;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.persistenceVersion;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.version;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.callType;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillAppCode;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillSuCode;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillBoId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillBeMetadataId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillBeId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillVoMetadataId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillVoId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillAppCode;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillSuCode;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillBoId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillBeMetadataId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillBeId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillVoMetadataId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillVoId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.isSupportAutomatinFlow;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.bizEventId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.bizActionId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.originId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.rootId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.parentId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.enabledParentEventRouterCond;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.enabledParentMappingCond;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.idPath;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.versionTag;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillBeMetadataRootId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillBeMetadataParentId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillBeMetadataIdPath;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillBeRootId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillBeParentId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillBeIdPath;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillVoMetadataRootId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillVoMetadataParentId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillVoMetadataIdPath;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillVoRootId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillVoParentId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.upBillVoIdPath;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillBeMetadataRootId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillBeMetadataParentId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillBeMetadataIdPath;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillBeRootId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillBeParentId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillBeIdPath;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillVoMetadataRootId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillVoMetadataParentId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillVoMetadataIdPath;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillVoRootId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillVoParentId;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.downBillVoIdPath;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.sysInit;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.origin;
    /** @type {?} */
    AbstractBillProcessEntity.prototype.colI18nValues;
}
