// @edp-bif/common-api/esm5/lib/entity/impl/AbstractBillCpAndWpEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/AbstractBillCpAndWpEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Entity, BpmnModelJsonHelper } from '../../ref';
var AbstractBillCpAndWpEntity = /** @class */ (function (_super) {
    tslib_1.__extends(AbstractBillCpAndWpEntity, _super);
    function AbstractBillCpAndWpEntity() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.sysInit = false;
        _this.isCreationProcess = false;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    AbstractBillCpAndWpEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.code = null;
        this.name = null;
        this.persistenceVersion = null;
        this.parentId = null;
        this.upBillVoId = null;
        this.downBillVoId = null;
        this.sysInit = false;
        this.isCreationProcess = false;
        this.originId = null;
        this.rootId = null;
        this.colI18nValues = null;
    };
    /**
     * @return {?}
     */
    AbstractBillCpAndWpEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "persistenceVersion", this.persistenceVersion);
        BpmnModelJsonHelper.SetAttributeValue(obj, "parentId", this.parentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoId", this.upBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoId", this.downBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sysInit", this.sysInit);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isCreationProcess", this.isCreationProcess);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originId", this.originId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "rootId", this.rootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "colI18nValues", this.colI18nValues);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AbstractBillCpAndWpEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.persistenceVersion = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "persistenceVersion");
        this.parentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "parentId");
        this.upBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoId");
        this.downBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoId");
        this.sysInit = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "sysInit");
        this.isCreationProcess = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isCreationProcess");
        this.originId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originId");
        this.rootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "rootId");
        this.colI18nValues = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "colI18nValues");
    };
    return AbstractBillCpAndWpEntity;
}(Entity));
export { AbstractBillCpAndWpEntity };
if (false) {
    /** @type {?} */
    AbstractBillCpAndWpEntity.prototype.code;
    /** @type {?} */
    AbstractBillCpAndWpEntity.prototype.name;
    /** @type {?} */
    AbstractBillCpAndWpEntity.prototype.persistenceVersion;
    /** @type {?} */
    AbstractBillCpAndWpEntity.prototype.parentId;
    /** @type {?} */
    AbstractBillCpAndWpEntity.prototype.upBillVoId;
    /** @type {?} */
    AbstractBillCpAndWpEntity.prototype.downBillVoId;
    /** @type {?} */
    AbstractBillCpAndWpEntity.prototype.sysInit;
    /** @type {?} */
    AbstractBillCpAndWpEntity.prototype.isCreationProcess;
    /** @type {?} */
    AbstractBillCpAndWpEntity.prototype.originId;
    /** @type {?} */
    AbstractBillCpAndWpEntity.prototype.rootId;
    /** @type {?} */
    AbstractBillCpAndWpEntity.prototype.colI18nValues;
}
