// @edp-bif/common-api/esm5/lib/entity/impl/BillTrackingSettings.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/BillTrackingSettings.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '../../ref';
import { BillTrackingDataShownMode } from '../BillTrackingDataShownMode';
import { BillTrackingNodeShownMode } from '../BillTrackingNodeShownMode';
var BillTrackingSettings = /** @class */ (function (_super) {
    tslib_1.__extends(BillTrackingSettings, _super);
    function BillTrackingSettings() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.billTrackingNodeShownMode = BillTrackingNodeShownMode.DEFAULT;
        _this.billTrackingDataShownMode = BillTrackingDataShownMode.DEFAULT;
        return _this;
    }
    ;
    /**
     * @protected
     * @return {?}
     */
    BillTrackingSettings.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.billTrackingNodeShownMode = BillTrackingNodeShownMode.DEFAULT;
        this.billNodeName = null;
        this.billNodeShownStyle = null;
        this.summaryFieldsInfo = null;
        this.billTrackingDataShownMode = BillTrackingDataShownMode.DEFAULT;
        this.customFieldContent = null;
        this.customFormContent = null;
        this.version = null;
    };
    /**
     * @return {?}
     */
    BillTrackingSettings.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billNodeName", this.billNodeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billTrackingNodeShownMode", this.billTrackingNodeShownMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billNodeShownStyle", this.billNodeShownStyle);
        BpmnModelJsonHelper.SetAttributeValue(obj, "summaryFieldsInfo", this.summaryFieldsInfo);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billTrackingDataShownMode", this.billTrackingDataShownMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "customFieldContent", this.customFieldContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "customFormContent", this.customFormContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "version", this.version);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BillTrackingSettings.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.billNodeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billNodeName");
        this.billTrackingNodeShownMode = BillTrackingDataShownMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billTrackingNodeShownMode")];
        this.billNodeShownStyle = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billNodeShownStyle");
        this.summaryFieldsInfo = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "summaryFieldsInfo");
        this.billTrackingDataShownMode = BillTrackingDataShownMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billTrackingDataShownMode")];
        this.customFieldContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "customFieldContent");
        this.customFormContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "customFormContent");
        this.version = (/** @type {?} */ (BpmnModelJsonHelper.GetAttributeValueAsString(obj, "version")));
    };
    return BillTrackingSettings;
}(JsonConvertableImpl));
export { BillTrackingSettings };
if (false) {
    /** @type {?} */
    BillTrackingSettings.prototype.billNodeName;
    /** @type {?} */
    BillTrackingSettings.prototype.billTrackingNodeShownMode;
    /** @type {?} */
    BillTrackingSettings.prototype.billNodeShownStyle;
    /** @type {?} */
    BillTrackingSettings.prototype.summaryFieldsInfo;
    /** @type {?} */
    BillTrackingSettings.prototype.billTrackingDataShownMode;
    /** @type {?} */
    BillTrackingSettings.prototype.customFieldContent;
    /** @type {?} */
    BillTrackingSettings.prototype.customFormContent;
    /**
     * 旧版本此字段为null，新版本为1.0
     * @type {?}
     */
    BillTrackingSettings.prototype.version;
    /* Skipping unhandled member: ;*/
}
