// @edp-bif/common-api/esm5/lib/entity/impl/BillTrackingConfigEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/BillTrackingConfigEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Entity, BpmnModelJsonHelper } from '../../ref';
var BillTrackingConfigEntity = /** @class */ (function (_super) {
    tslib_1.__extends(BillTrackingConfigEntity, _super);
    function BillTrackingConfigEntity() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isEnabled = false;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    BillTrackingConfigEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.suCode = null;
        this.classTypeName = null;
        this.isEnabled = false;
    };
    /**
     * @return {?}
     */
    BillTrackingConfigEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "classTypeName", this.classTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BillTrackingConfigEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.classTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "classTypeName");
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
    };
    return BillTrackingConfigEntity;
}(Entity));
export { BillTrackingConfigEntity };
if (false) {
    /** @type {?} */
    BillTrackingConfigEntity.prototype.suCode;
    /** @type {?} */
    BillTrackingConfigEntity.prototype.classTypeName;
    /** @type {?} */
    BillTrackingConfigEntity.prototype.isEnabled;
}
