// @edp-bif/common-api/esm5/lib/entity/impl/ComponentConditionContent.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ComponentConditionContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from "../../ref";
import { ConditionContent } from './ConditionContent';
var ComponentConditionContent = /** @class */ (function (_super) {
    tslib_1.__extends(ComponentConditionContent, _super);
    function ComponentConditionContent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    ComponentConditionContent.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.componentMetadataId = null;
        this.componentId = null;
        this.name = null;
    };
    /**
     * @return {?}
     */
    ComponentConditionContent.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentMetadataId", this.componentMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentId", this.componentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ComponentConditionContent.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.componentMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentMetadataId");
        this.componentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentId");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
    };
    return ComponentConditionContent;
}(ConditionContent));
export { ComponentConditionContent };
if (false) {
    /** @type {?} */
    ComponentConditionContent.prototype.componentMetadataId;
    /** @type {?} */
    ComponentConditionContent.prototype.componentId;
    /** @type {?} */
    ComponentConditionContent.prototype.name;
}
