// @edp-bif/common-api/esm5/lib/entity/impl/DocumentAssociationEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/DocumentAssociationEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, Entity } from "@edp-aif/common-api";
import { AssociationType } from "./AssociationType";
var DocumentAssociationEntity = /** @class */ (function (_super) {
    tslib_1.__extends(DocumentAssociationEntity, _super);
    function DocumentAssociationEntity() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.associationType = AssociationType.DEFAULT;
        _this.sequenceNum = 0;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    DocumentAssociationEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.associationType = AssociationType.DEFAULT;
        this.sequenceNum = 0;
        this.originDocuSucode = null;
        this.originDocuVoId = null;
        this.originDocuTypeId = null;
        this.originDocuDataId = null;
        this.originDocuItemId = null;
        this.originDocuItemCode = null;
        this.originDocuDetailId = null;
        this.originDocuDataIdPath = null;
        this.upDocuSucode = null;
        this.upDocuVoId = null;
        this.upDocuTypeId = null;
        this.upDocuDataId = null;
        this.upDocuItemId = null;
        this.upDocuDetailId = null;
        this.upDocuDataIdPath = null;
        this.upDocuCode = null;
        this.upDocuDate = null;
        this.upDocuAbstract = null;
        this.downDocuSucode = null;
        this.downDocuVoId = null;
        this.downDocuTypeId = null;
        this.downDocuDataId = null;
        this.downDocuItemId = null;
        this.downDocuDetailId = null;
        this.downDocuDataIdPath = null;
        this.downDocuCode = null;
        this.downDocuDate = null;
        this.downDocuAbstract = null;
    };
    /**
     * @return {?}
     */
    DocumentAssociationEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "associationType", this.associationType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sequenceNum", this.sequenceNum);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuSucode", this.originDocuSucode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuVoId", this.originDocuVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuTypeId", this.originDocuTypeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuDataId", this.originDocuDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuItemId", this.originDocuItemId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuItemCode", this.originDocuItemCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuDetailId", this.originDocuDetailId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuDataIdPath", this.originDocuDataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuSucode", this.upDocuSucode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuVoId", this.upDocuVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuTypeId", this.upDocuTypeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDataId", this.upDocuDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuItemId", this.upDocuItemId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDetailId", this.upDocuDetailId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDataIdPath", this.upDocuDataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuCode", this.upDocuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDate", this.upDocuDate);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuAbstract", this.upDocuAbstract);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuSucode", this.downDocuSucode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuVoId", this.downDocuVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuTypeId", this.downDocuTypeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDataId", this.downDocuDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuItemId", this.downDocuItemId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDetailId", this.downDocuDetailId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDataIdPath", this.downDocuDataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuCode", this.downDocuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDate", this.downDocuDate);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuAbstract", this.downDocuAbstract);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    DocumentAssociationEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.associationType = AssociationType[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "associationType")];
        this.sequenceNum = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "sequenceNum");
        this.originDocuSucode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuSucode");
        this.originDocuVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuVoId");
        this.originDocuTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuTypeId");
        this.originDocuDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuDataId");
        this.originDocuItemId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuItemId");
        this.originDocuItemCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuItemCode");
        this.originDocuDetailId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuDetailId");
        this.originDocuDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuDataIdPath");
        this.upDocuSucode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuSucode");
        this.upDocuVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuVoId");
        this.upDocuTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuTypeId");
        this.upDocuDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDataId");
        this.upDocuItemId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuItemId");
        this.upDocuDetailId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDetailId");
        this.upDocuDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDataIdPath");
        this.upDocuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuCode");
        this.upDocuDate = BpmnModelJsonHelper.GetAttributeValueAsDateTime(obj, "upDocuDate");
        this.upDocuAbstract = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuAbstract");
        this.downDocuSucode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuSucode");
        this.downDocuVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuVoId");
        this.downDocuTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuTypeId");
        this.downDocuDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDataId");
        this.downDocuItemId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuItemId");
        this.downDocuDetailId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDetailId");
        this.downDocuDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDataIdPath");
        this.downDocuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuCode");
        this.downDocuDate = BpmnModelJsonHelper.GetAttributeValueAsDateTime(obj, "downDocuDate");
        this.downDocuAbstract = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuAbstract");
    };
    return DocumentAssociationEntity;
}(Entity));
export { DocumentAssociationEntity };
if (false) {
    /** @type {?} */
    DocumentAssociationEntity.prototype.associationType;
    /** @type {?} */
    DocumentAssociationEntity.prototype.sequenceNum;
    /** @type {?} */
    DocumentAssociationEntity.prototype.originDocuSucode;
    /** @type {?} */
    DocumentAssociationEntity.prototype.originDocuVoId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.originDocuTypeId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.originDocuDataId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.originDocuItemId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.originDocuItemCode;
    /** @type {?} */
    DocumentAssociationEntity.prototype.originDocuDetailId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.originDocuDataIdPath;
    /** @type {?} */
    DocumentAssociationEntity.prototype.upDocuSucode;
    /** @type {?} */
    DocumentAssociationEntity.prototype.upDocuVoId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.upDocuTypeId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.upDocuDataId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.upDocuItemId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.upDocuDetailId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.upDocuDataIdPath;
    /** @type {?} */
    DocumentAssociationEntity.prototype.upDocuCode;
    /** @type {?} */
    DocumentAssociationEntity.prototype.upDocuDate;
    /** @type {?} */
    DocumentAssociationEntity.prototype.upDocuAbstract;
    /** @type {?} */
    DocumentAssociationEntity.prototype.downDocuSucode;
    /** @type {?} */
    DocumentAssociationEntity.prototype.downDocuVoId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.downDocuTypeId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.downDocuDataId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.downDocuItemId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.downDocuDetailId;
    /** @type {?} */
    DocumentAssociationEntity.prototype.downDocuDataIdPath;
    /** @type {?} */
    DocumentAssociationEntity.prototype.downDocuCode;
    /** @type {?} */
    DocumentAssociationEntity.prototype.downDocuDate;
    /** @type {?} */
    DocumentAssociationEntity.prototype.downDocuAbstract;
}
