// @edp-bif/common-api/esm5/lib/entity/impl/DtBillWritebackProcessEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/DtBillWritebackProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '../../ref';
import { AbstractBillWritebackProcessEntity } from './AbstractBillWritebackProcessEntity';
var DtBillWritebackProcessEntity = /** @class */ (function (_super) {
    tslib_1.__extends(DtBillWritebackProcessEntity, _super);
    function DtBillWritebackProcessEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    DtBillWritebackProcessEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.eventRouterConditionsContent = null;
        this.mappingConditionsContent = null;
        this.mappingModeContent = null;
        this.saveModeContent = null;
    };
    /**
     * @return {?}
     */
    DtBillWritebackProcessEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsContent", this.eventRouterConditionsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsContent", this.mappingConditionsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeContent", this.mappingModeContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeContent", this.saveModeContent);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    DtBillWritebackProcessEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.eventRouterConditionsContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsContent");
        this.mappingConditionsContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsContent");
        this.mappingModeContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeContent");
        this.saveModeContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeContent");
    };
    return DtBillWritebackProcessEntity;
}(AbstractBillWritebackProcessEntity));
export { DtBillWritebackProcessEntity };
if (false) {
    /** @type {?} */
    DtBillWritebackProcessEntity.prototype.eventRouterConditionsContent;
    /** @type {?} */
    DtBillWritebackProcessEntity.prototype.mappingConditionsContent;
    /** @type {?} */
    DtBillWritebackProcessEntity.prototype.mappingModeContent;
    /** @type {?} */
    DtBillWritebackProcessEntity.prototype.saveModeContent;
}
