// @edp-bif/common-api/esm5/lib/entity/impl/MappingRuleContent.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/MappingRuleContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from '../../ref';
var MappingRuleContent = /** @class */ (function (_super) {
    tslib_1.__extends(MappingRuleContent, _super);
    function MappingRuleContent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    MappingRuleContent.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.mappingRuleMetadataId = null;
        this.mappingRuleId = null;
        this.mappingContent = null;
        this.mappingConfig = null;
    };
    /**
     * @return {?}
     */
    MappingRuleContent.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleMetadataId", this.mappingRuleMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleId", this.mappingRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingContent", this.mappingContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConfig", this.mappingConfig);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    MappingRuleContent.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.mappingRuleMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleMetadataId");
        this.mappingRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleId");
        this.mappingContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingContent");
        this.mappingConfig = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConfig");
    };
    return MappingRuleContent;
}(JsonConvertableImpl));
export { MappingRuleContent };
if (false) {
    /** @type {?} */
    MappingRuleContent.prototype.mappingRuleMetadataId;
    /** @type {?} */
    MappingRuleContent.prototype.mappingRuleId;
    /** @type {?} */
    MappingRuleContent.prototype.mappingContent;
    /** @type {?} */
    MappingRuleContent.prototype.mappingConfig;
}
