// @edp-bif/common-api/esm5/lib/entity/impl/MappingModeEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/MappingModeEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from "../../ref";
import { MappingMode } from '../MappingMode';
/**
 * 映射方式实体类。
 */
var /**
 * 映射方式实体类。
 */
MappingModeEntity = /** @class */ (function (_super) {
    tslib_1.__extends(MappingModeEntity, _super);
    function MappingModeEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    MappingModeEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.mappingMode = null;
        this.content = null;
        this.mappingRuleContent = null;
        this.componentContent = null;
        this.voActionContent = null;
        this.expressionContent = null;
    };
    /**
     * @return {?}
     */
    MappingModeEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingMode", this.mappingMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleContent", this.mappingRuleContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentContent", this.componentContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voActionContent", this.voActionContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "expressionContent", this.expressionContent);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    MappingModeEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.mappingMode = MappingMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingMode")];
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
        this.mappingRuleContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleContent");
        this.componentContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentContent");
        this.voActionContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voActionContent");
        this.expressionContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "expressionContent");
    };
    return MappingModeEntity;
}(JsonConvertableImpl));
/**
 * 映射方式实体类。
 */
export { MappingModeEntity };
if (false) {
    /** @type {?} */
    MappingModeEntity.prototype.mappingMode;
    /** @type {?} */
    MappingModeEntity.prototype.content;
    /** @type {?} */
    MappingModeEntity.prototype.mappingRuleContent;
    /** @type {?} */
    MappingModeEntity.prototype.componentContent;
    /** @type {?} */
    MappingModeEntity.prototype.voActionContent;
    /** @type {?} */
    MappingModeEntity.prototype.expressionContent;
}
