// @edp-bif/common-api/esm5/lib/entity/impl/RtBillWritebackProcessEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/RtBillWritebackProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '../../ref';
import { AbstractBillWritebackProcessEntity } from './AbstractBillWritebackProcessEntity';
var RtBillWritebackProcessEntity = /** @class */ (function (_super) {
    tslib_1.__extends(RtBillWritebackProcessEntity, _super);
    function RtBillWritebackProcessEntity() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isEnabled = false;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RtBillWritebackProcessEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.isEnabled = false;
        this.eventRouterConditionsId = null;
        this.mappingConditionsId = null;
        this.mappingModeId = null;
        this.saveModeId = null;
        this.dtBillWritebackProcessId = null;
    };
    /**
     * @return {?}
     */
    RtBillWritebackProcessEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsId", this.eventRouterConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsId", this.mappingConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeId", this.mappingModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeId", this.saveModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dtBillWritebackProcessId", this.dtBillWritebackProcessId);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RtBillWritebackProcessEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.eventRouterConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsId");
        this.mappingConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsId");
        this.mappingModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeId");
        this.saveModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeId");
        this.dtBillWritebackProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dtBillWritebackProcessId");
    };
    return RtBillWritebackProcessEntity;
}(AbstractBillWritebackProcessEntity));
export { RtBillWritebackProcessEntity };
if (false) {
    /** @type {?} */
    RtBillWritebackProcessEntity.prototype.isEnabled;
    /** @type {?} */
    RtBillWritebackProcessEntity.prototype.eventRouterConditionsId;
    /** @type {?} */
    RtBillWritebackProcessEntity.prototype.mappingConditionsId;
    /** @type {?} */
    RtBillWritebackProcessEntity.prototype.mappingModeId;
    /** @type {?} */
    RtBillWritebackProcessEntity.prototype.saveModeId;
    /** @type {?} */
    RtBillWritebackProcessEntity.prototype.dtBillWritebackProcessId;
}
