// @edp-bif/common-api/esm5/lib/entity/impl/RtBillCreationProcessEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/RtBillCreationProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '../../ref';
import { AbstractBillCreationProcessEntity } from './AbstractBillCreationProcessEntity';
var RtBillCreationProcessEntity = /** @class */ (function (_super) {
    tslib_1.__extends(RtBillCreationProcessEntity, _super);
    function RtBillCreationProcessEntity() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isEnabled = false;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RtBillCreationProcessEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.isEnabled = false;
        this.eventRouterConditionsId = null;
        this.mappingConditionsId = null;
        this.mappingModeId = null;
        this.saveModeId = null;
        this.formSettingsId = null;
        this.dtBillCreationProcessId = null;
        this.upBillTrackingSettingsId = null;
        this.downBillTrackingSettingsId = null;
        this.reverseWriteRuleId = null;
        this.taskCenterConfig4Push = null;
        this.taskCenterConfig4Pull = null;
        this.isSupportPull = false;
        this.isSupportPush = false;
    };
    /**
     * @return {?}
     */
    RtBillCreationProcessEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsId", this.eventRouterConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsId", this.mappingConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeId", this.mappingModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeId", this.saveModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "formSettingsId", this.formSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dtBillCreationProcessId", this.dtBillCreationProcessId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettingsId", this.upBillTrackingSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettingsId", this.downBillTrackingSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleId", this.reverseWriteRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Push", this.taskCenterConfig4Push);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Pull", this.taskCenterConfig4Pull);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPull", this.isSupportPull);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPush", this.isSupportPush);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RtBillCreationProcessEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.eventRouterConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsId");
        this.mappingConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsId");
        this.mappingModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeId");
        this.saveModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeId");
        this.formSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettingsId");
        this.dtBillCreationProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dtBillCreationProcessId");
        this.upBillTrackingSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettingsId");
        this.downBillTrackingSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettingsId");
        this.reverseWriteRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleId");
        this.taskCenterConfig4Push = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Push");
        this.taskCenterConfig4Pull = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Pull");
        this.isSupportPull = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPull");
        this.isSupportPush = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPush");
    };
    return RtBillCreationProcessEntity;
}(AbstractBillCreationProcessEntity));
export { RtBillCreationProcessEntity };
if (false) {
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.isEnabled;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.eventRouterConditionsId;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.mappingConditionsId;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.mappingModeId;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.saveModeId;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.formSettingsId;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.dtBillCreationProcessId;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.upBillTrackingSettingsId;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.downBillTrackingSettingsId;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.reverseWriteRuleId;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.taskCenterConfig4Push;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.taskCenterConfig4Pull;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.isSupportPull;
    /** @type {?} */
    RtBillCreationProcessEntity.prototype.isSupportPush;
}
