// @edp-bif/common-api/esm5/lib/entity/impl/SaveModeEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/SaveModeEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '../../ref';
import { SaveMode } from '../SaveMode';
/**
 * 保存模式实体类。
 */
var /**
 * 保存模式实体类。
 */
SaveModeEntity = /** @class */ (function (_super) {
    tslib_1.__extends(SaveModeEntity, _super);
    function SaveModeEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    SaveModeEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.saveMode = null;
        this.content = null;
        this.componentContent = null;
        this.voActionContent = null;
        this.taskCenterContent = null;
    };
    /**
     * @return {?}
     */
    SaveModeEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveMode", this.saveMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentContent", this.componentContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voActionContent", this.voActionContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterContent", this.taskCenterContent);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    SaveModeEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.saveMode = SaveMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveMode")];
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
        this.componentContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentContent");
        this.voActionContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voActionContent");
        this.taskCenterContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterContent");
    };
    return SaveModeEntity;
}(JsonConvertableImpl));
/**
 * 保存模式实体类。
 */
export { SaveModeEntity };
if (false) {
    /** @type {?} */
    SaveModeEntity.prototype.saveMode;
    /** @type {?} */
    SaveModeEntity.prototype.content;
    /** @type {?} */
    SaveModeEntity.prototype.componentContent;
    /** @type {?} */
    SaveModeEntity.prototype.voActionContent;
    /** @type {?} */
    SaveModeEntity.prototype.taskCenterContent;
}
