// @edp-bif/common-api/esm5/lib/entity/impl/TableSchema.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/TableSchema.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from "@edp-aif/common-api";
var TableSchema = /** @class */ (function (_super) {
    tslib_1.__extends(TableSchema, _super);
    function TableSchema() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.maxDisplayNumber = 2000;
        _this.isMainTable = false;
        _this.isDisplay = true;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    TableSchema.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.code = null;
        this.maxDisplayNumber = 2000;
        this.isMainTable = false;
        this.isDisplay = true;
    };
    /**
     * @return {?}
     */
    TableSchema.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "maxDisplayNumber", this.maxDisplayNumber);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isMainTable", this.isMainTable);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isDisplay", this.isDisplay);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    TableSchema.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.maxDisplayNumber = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "maxDisplayNumber");
        this.isMainTable = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isMainTable");
        this.isDisplay = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isDisplay");
    };
    return TableSchema;
}(JsonConvertableImpl));
export { TableSchema };
if (false) {
    /** @type {?} */
    TableSchema.prototype.code;
    /** @type {?} */
    TableSchema.prototype.maxDisplayNumber;
    /** @type {?} */
    TableSchema.prototype.isMainTable;
    /** @type {?} */
    TableSchema.prototype.isDisplay;
}
