// @edp-bif/common-api/esm5/lib/entity/impl/ThirdSchemaInfoEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ThirdSchemaInfoEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, Entity } from "@edp-aif/common-api";
var ThirdSchemaInfoEntity = /** @class */ (function (_super) {
    tslib_1.__extends(ThirdSchemaInfoEntity, _super);
    function ThirdSchemaInfoEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    ThirdSchemaInfoEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.schemaCategoryId = null;
        this.schemaCategoryName = null;
        this.schemaConvertorClassTypeName = null;
        this.trackingConfigClassTypeName = null;
        this.suCode = null;
    };
    /**
     * @return {?}
     */
    ThirdSchemaInfoEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaCategoryId", this.schemaCategoryId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaCategoryName", this.schemaCategoryName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaConvertorClassTypeName", this.schemaConvertorClassTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "trackingConfigClassTypeName", this.trackingConfigClassTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ThirdSchemaInfoEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.schemaCategoryId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaCategoryId");
        this.schemaCategoryName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaCategoryName");
        this.schemaConvertorClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaConvertorClassTypeName");
        this.trackingConfigClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "trackingConfigClassTypeName");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
    };
    return ThirdSchemaInfoEntity;
}(Entity));
export { ThirdSchemaInfoEntity };
if (false) {
    /** @type {?} */
    ThirdSchemaInfoEntity.prototype.schemaCategoryId;
    /** @type {?} */
    ThirdSchemaInfoEntity.prototype.schemaCategoryName;
    /** @type {?} */
    ThirdSchemaInfoEntity.prototype.schemaConvertorClassTypeName;
    /** @type {?} */
    ThirdSchemaInfoEntity.prototype.trackingConfigClassTypeName;
    /** @type {?} */
    ThirdSchemaInfoEntity.prototype.suCode;
}
