// @gsp-lcm/bo-dt-service/bundles/gsp-lcm-bo-dt-service.umd.js
(function (global, factory) {
   typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('rxjs/observable/of'), require('@angular/common/http'), require('@angular/core')) :
   typeof define === 'function' && define.amd ? define('@gsp-lcm/bo-dt-service', ['exports', 'rxjs/observable/of', '@angular/common/http', '@angular/core'], factory) :
   (factory((global['gsp-lcm'] = global['gsp-lcm'] || {}, global['gsp-lcm']['bo-dt-service'] = {}),global.rxjs['observable/of'],global.ng.common.http,global.ng.core));
}(this, (function (exports,of,http,core) { 'use strict';

   /**
    * @fileoverview added by tsickle
    * Generated from: lib/business-object-dt.service.ts
    * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
    */
   /** @type {?} */
   var PROJECTS_DIRECTORY = '/api/dev/main/v1.0/business-object-trees';
   /** @type {?} */
   var url = '/api/runtime/sys/v1.0/business-objects/';
   var BusinessObjectDtService = /** @class */ (function () {
       function BusinessObjectDtService(http$$1) {
           this.http = http$$1;
       }
       /**
        * @return {?}
        */
       BusinessObjectDtService.prototype.getBusinessObjectTree = /**
        * @return {?}
        */
           function () {
               /** @type {?} */
               var http$$1 = this.http;
               /** @type {?} */
               var dboUrl = PROJECTS_DIRECTORY;
               return http$$1.get(dboUrl).pipe();
           };
       /**
        * @param {?} boId
        * @return {?}
        */
       BusinessObjectDtService.prototype.getBusinessObject = /**
        * @param {?} boId
        * @return {?}
        */
           function (boId) {
               /** @type {?} */
               var getUrl = url + boId;
               return this.http.get(getUrl).pipe();
           };
       /**
        * @return {?}
        */
       BusinessObjectDtService.prototype.getBusinessObjects = /**
        * @return {?}
        */
           function () {
               return this.http.get(url).pipe();
           };
       /**
        * @param {?} productId
        * @param {?} layer
        * @return {?}
        */
       BusinessObjectDtService.prototype.getBoListByProductIdAndLayer = /**
        * @param {?} productId
        * @param {?} layer
        * @return {?}
        */
           function (productId, layer) {
               /** @type {?} */
               var getUrl = ' /api/runtime/sys/v1.0/business-objects?productId=' + productId + '&layer=' + layer;
               return this.http.get(getUrl).pipe();
           };
       /**
      * 保存
      */
       /**
        * 保存
        * @param {?} postData
        * @return {?}
        */
       BusinessObjectDtService.prototype.save = /**
        * 保存
        * @param {?} postData
        * @return {?}
        */
           function (postData) {
               return this.http.post(url, postData).pipe();
           };
       /**
        * @param {?} postData
        * @return {?}
        */
       BusinessObjectDtService.prototype.saveChanges = /**
        * @param {?} postData
        * @return {?}
        */
           function (postData) {
               return this.http.put(url, postData).pipe();
           };
       /**
          * 删除并重新加载
          */
       /**
        * 删除并重新加载
        * @param {?} dataId
        * @return {?}
        */
       BusinessObjectDtService.prototype.remove = /**
        * 删除并重新加载
        * @param {?} dataId
        * @return {?}
        */
           function (dataId) {
               /** @type {?} */
               var deleteUrl = url + dataId;
               return this.http.delete(deleteUrl, null).pipe();
           };
       /**
        * @param {?} boId
        * @return {?}
        */
       BusinessObjectDtService.prototype.getDevBasicBOInfo = /**
        * @param {?} boId
        * @return {?}
        */
           function (boId) {
               /** @type {?} */
               var devUrl = '/api/runtime/sys/v1.0/business-objects/dev-basic-boinfos/';
               /** @type {?} */
               var getUrl = devUrl + boId;
               return this.http.get(getUrl).pipe();
           };
       /**
        * @param {?} parentId
        * @return {?}
        */
       BusinessObjectDtService.prototype.getBOListByParentId = /**
        * @param {?} parentId
        * @return {?}
        */
           function (parentId) {
               /** @type {?} */
               var getUrl = url + '?parentID=' + parentId;
               return this.http.get(getUrl).pipe();
           };
       /**
        * @private
        * @template T
        * @param {?=} operation
        * @param {?=} result
        * @return {?}
        */
       BusinessObjectDtService.prototype.handleError = /**
        * @private
        * @template T
        * @param {?=} operation
        * @param {?=} result
        * @return {?}
        */
           function (operation, result) {
               if (operation === void 0) {
                   operation = 'operation';
               }
               return ( /**
                * @param {?} error
                * @return {?}
                */function (error) {
                   console.error(error); // log to console instead
                   return of.of(( /** @type {?} */(result)));
               });
           };
       /**
        * @param {?} parentId
        * @return {?}
        */
       BusinessObjectDtService.prototype.getBusinessObjectsByParentId = /**
        * @param {?} parentId
        * @return {?}
        */
           function (parentId) {
               /** @type {?} */
               var headers = new http.HttpHeaders().set('sessionId', this.sessionId);
               return this.http.get(url + parentId + '/children').pipe();
           };
       BusinessObjectDtService.decorators = [
           { type: core.Injectable }
       ];
       /** @nocollapse */
       BusinessObjectDtService.ctorParameters = function () {
           return [
               { type: http.HttpClient }
           ];
       };
       return BusinessObjectDtService;
   }());

   /**
    * @fileoverview added by tsickle
    * Generated from: lib/bo-dt-service.module.ts
    * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
    */
   var BoDtServiceModule = /** @class */ (function () {
       function BoDtServiceModule() {
       }
       BoDtServiceModule.decorators = [
           { type: core.NgModule, args: [{
                       declarations: [],
                       imports: [],
                       exports: [],
                       providers: [BusinessObjectDtService]
                   },] }
       ];
       return BoDtServiceModule;
   }());

   /**
    * @fileoverview added by tsickle
    * Generated from: lib/businessobjectentity.ts
    * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
    */
   /**
    * 业务对象实体
    */
   var /**
    * 业务对象实体
    */ BusinessObjectEntity = /** @class */ (function () {
       function BusinessObjectEntity() {
           this.code = '';
           this.name = '';
       }
       return BusinessObjectEntity;
   }());

   /**
    * @fileoverview added by tsickle
    * Generated from: public-api.ts
    * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
    */

   /**
    * @fileoverview added by tsickle
    * Generated from: gsp-lcm-bo-dt-service.ts
    * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
    */

   exports.BusinessObjectDtService = BusinessObjectDtService;
   exports.BoDtServiceModule = BoDtServiceModule;
   exports.BusinessObjectEntity = BusinessObjectEntity;

   Object.defineProperty(exports, '__esModule', { value: true });

})));
