// @gsp-wf/wf-process-model/esm5/lib/model/parameter/FormalParameter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelHelper, MexElement, DataType } from "@edp-pmf/bpmn-model";
import { ParameterMode } from "./ParameterMode";
var FormalParameter = /** @class */ (function (_super) {
    tslib_1.__extends(FormalParameter, _super);
    function FormalParameter(model, id, code, name, desc, index, datatype, mode) {
        var _this = _super.call(this, model) || this;
        _this.Id = id || '';
        _this.Code = code || '';
        _this.Name = name || '';
        _this.Description = desc || '';
        _this.index = index || 0;
        _this.DataType = datatype || new DataType(model);
        _this.Mode = mode || ParameterMode.IN;
        return _this;
    }
    Object.defineProperty(FormalParameter.prototype, "Index", {
        get: /**
         * @return {?}
         */
        function () {
            return this.index;
        },
        set: /**
         * @param {?} index
         * @return {?}
         */
        function (index) {
            this.index = index;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @protected
     * @return {?}
     */
    FormalParameter.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    FormalParameter.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "Id", this.Id);
        BpmnModelHelper.SetAttributeValue(obj, "Code", this.Code);
        BpmnModelHelper.SetAttributeValue(obj, "Name", this.Name);
        BpmnModelHelper.SetAttributeValue(obj, "Description", this.Description);
        BpmnModelHelper.SetAttributeValue(obj, "index", this.index);
        BpmnModelHelper.SetAttributeValue(obj, "DataType", this.DataType.ConvertToJsonObject());
        BpmnModelHelper.SetAttributeValue(obj, "Mode", this.Mode);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormalParameter.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Id = obj.Id;
        this.Name = obj.Name;
        this.index = obj.index;
        this.Description = obj.Description;
        this.Code = obj.Code;
        /** @type {?} */
        var type = new DataType(this.Model);
        type.LoadFromJsonObject(obj.DataType);
        this.DataType = type;
        this.Mode = ((/** @type {?} */ (obj.Mode)));
    };
    /**
     * @return {?}
     */
    FormalParameter.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var f = new FormalParameter(this.Model);
        return f;
    };
    return FormalParameter;
}(MexElement));
export { FormalParameter };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FormalParameter.prototype.index;
    /** @type {?} */
    FormalParameter.prototype.Id;
    /** @type {?} */
    FormalParameter.prototype.Code;
    /** @type {?} */
    FormalParameter.prototype.Name;
    /** @type {?} */
    FormalParameter.prototype.Description;
    /** @type {?} */
    FormalParameter.prototype.DataType;
    /** @type {?} */
    FormalParameter.prototype.Mode;
}
