// @gsp-wf/wf-process-model/esm5/lib/model/parameter/WFActualParameter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, MexExtensionElement } from "@edp-pmf/bpmn-model";
var ActualParameter = /** @class */ (function (_super) {
    tslib_1.__extends(ActualParameter, _super);
    function ActualParameter(model, name, value, valueText, tag) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = ActualParameter.Clr_Type_ID;
        _this.Name = name;
        _this.Value = value;
        _this.ValueText = valueText;
        _this.Tag = tag;
        return _this;
    }
    /**
     * @return {?}
     */
    ActualParameter.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new ActualParameter(this.Model);
    };
    /**
     * @protected
     * @return {?}
     */
    ActualParameter.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    ActualParameter.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "Name", this.Name);
        BpmnModelHelper.SetAttributeValue(obj, "Value", this.Value);
        BpmnModelHelper.SetAttributeValue(obj, "ValueText", this.ValueText);
        BpmnModelHelper.SetAttributeValue(obj, "Tag", this.Tag);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ActualParameter.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Name = obj.Name;
        this.Value = obj.Value;
        this.ValueText = obj.ValueText;
        this.Tag = obj.Tag;
    };
    /**
     * @return {?}
     */
    ActualParameter.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var parameter = new ActualParameter(this.Model);
        parameter.Name = this.Name;
        parameter.Value = this.Value;
        parameter.Tag = this.Tag;
        parameter.ValueText = this.ValueText;
        return parameter;
    };
    ActualParameter.Clr_Type_ID = "wf.Mex.ActualParameter";
    return ActualParameter;
}(MexExtensionElement));
export { ActualParameter };
if (false) {
    /** @type {?} */
    ActualParameter.Clr_Type_ID;
    /** @type {?} */
    ActualParameter.prototype.ClrTypeID;
    /** @type {?} */
    ActualParameter.prototype.Name;
    /** @type {?} */
    ActualParameter.prototype.Value;
    /** @type {?} */
    ActualParameter.prototype.ValueText;
    /** @type {?} */
    ActualParameter.prototype.Tag;
}
BpmnModelHelper.RegisterElementType(new ActualParameter(new BpmnModel()));
