// @progress/kendo-angular-dialog/dist/es/window/draggable.directive.js
import { EventEmitter, ElementRef, Directive, NgZone } from '@angular/core';
import Draggable from '@telerik/kendo-draggable';
/**
 * @hidden
 */
var DraggableDirective = /** @class */ (function () {
    function DraggableDirective(elem, ngZone) {
        this.elem = elem;
        this.ngZone = ngZone;
        this.kendo = {
            drag: new EventEmitter(),
            press: new EventEmitter(),
            release: new EventEmitter()
        };
    }
    DraggableDirective.prototype.ngOnInit = function () {
        var _this = this;
        if (typeof document !== 'undefined') {
            this.draggable = new Draggable({
                drag: function (e) { return _this.kendo.drag.next(e); },
                press: function (e) { return _this.kendo.press.next(e); },
                release: function (e) { return _this.kendo.release.next(e); }
            });
            this.ngZone.runOutsideAngular(function () { return _this.draggable.bindTo(_this.elem.nativeElement); });
        }
    };
    DraggableDirective.prototype.ngOnDestroy = function () {
        if (typeof document !== 'undefined') {
            this.draggable.destroy();
        }
    };
    Object.defineProperty(DraggableDirective.prototype, "nativeElement", {
        get: function () {
            return this.elem.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    DraggableDirective.decorators = [
        { type: Directive, args: [{
                    outputs: ['kendo.press', 'kendo.drag', 'kendo.release'],
                    selector: '[kendoWindowDraggable]'
                },] },
    ];
    /** @nocollapse */
    DraggableDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: NgZone, },
    ]; };
    return DraggableDirective;
}());
export { DraggableDirective };
