// @gsp-cmp/ubcc-param-data-ui/esm5/lib/ubcc-param-data-ui-root.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ubcc-param-data-ui-root.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { ButtonsModule } from '@progress/kendo-angular-buttons';
import { FrameworkService, FrmEventBus, FuncsService, LoadingService } from '@gsp-sys/rtf-common';
import { HttpService, SessionService, CacheService } from '@ecp-caf/caf-common';
import { NotifyService, NotifyModule } from '@farris/ui-notify';
import { InputGroupModule } from '@farris/ui-input-group';
import { HttpClientModule } from '@angular/common/http';
import { LoadingModule } from '@farris/ui-loading';
import { ModalModule, BsModalService } from '@farris/ui-modal';
import { MessagerModule } from '@farris/ui-messager';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { ComboListModule } from '@farris/ui-combo-list';
import { LookupModule } from '@farris/ui-lookup';
import { FarrisButtonModule } from '@farris/ui-button';
import { FarrisDatePickerModule } from '@farris/ui-datepicker';
import { FarrisSectionModule } from '@farris/ui-section';
import { UbccParamDataUiComponent } from './ubcc-param-data-ui/ubcc-param-data-ui.component';
import { UbccParamDataManangerService } from './service/http-service';
import { NumberSpinnerModule } from '@farris/ui-number-spinner';
import { GSPMetadataRTServiceModule } from '@gsp-lcm/metadatart-selector';
import { UbccParamDataUiDialogComponent } from './ubcc-param-data-ui-dialog/ubcc-param-data-ui-dialog.component';
import { UbccParamDataUiRootComponent } from './ubcc-param-data-ui-root.component';
import { DimensionLookupControlComponent } from './controls/dimension-control/dimension-lookup-control/dimension-lookup-control.component';
import { DimensionDropdownlistControlComponent } from './controls/dimension-control/dimension-dropdownlist-control/dimension-dropdownlist-control.component';
import { DimensionInputControlComponent } from './controls/dimension-control/dimension-input-control/dimension-input-control.component';
import { CheckboxControlComponent } from './controls/param-control/checkbox-control/checkbox-control.component';
import { InputControlComponent } from './controls/param-control/input-control/input-control.component';
import { DateControlComponent } from './controls/param-control/date-control/date-control.component';
import { DropdownlistControlComponent } from './controls/param-control/dropdownlist-control/dropdownlist-control.component';
import { LookupControlComponent } from './controls/param-control/lookup-control/lookup-control.component';
import { NumberControlComponent } from './controls/param-control/number-control/number-control.component';
import { UbccParamDataUiService } from './ubcc-param-data-ui.service';
import { UbccParamDataUtils } from './utils/param-data-utils';
import { UbccParamDimensionUiComponent } from './ubcc-param-dimension-ui/ubcc-param-dimension-ui.component';
import { UbccParamDimensionUiRootComponent } from './ubcc-param-dimension-ui-root/ubcc-param-dimension-ui-root.component';
import { FarrisTooltipModule } from '@farris/ui-tooltip';
import { FarrisPanelModule } from '@farris/ui-panel';
var UbccParamDataUiRootModule = /** @class */ (function () {
    function UbccParamDataUiRootModule() {
    }
    UbccParamDataUiRootModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        UbccParamDataUiComponent,
                        InputControlComponent,
                        CheckboxControlComponent,
                        DateControlComponent,
                        DropdownlistControlComponent,
                        LookupControlComponent,
                        NumberControlComponent,
                        UbccParamDataUiDialogComponent,
                        UbccParamDataUiRootComponent,
                        DimensionLookupControlComponent,
                        DimensionDropdownlistControlComponent,
                        DimensionInputControlComponent,
                        UbccParamDimensionUiComponent,
                        UbccParamDimensionUiRootComponent
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        ButtonsModule,
                        InputGroupModule,
                        HttpClientModule,
                        LoadingModule.forRoot({
                            message: '请稍候...'
                        }),
                        NotifyModule.forRoot(),
                        ModalModule.forRoot(),
                        MessagerModule.forRoot(),
                        FarrisDialogModule,
                        LookupModule,
                        FarrisButtonModule,
                        FarrisDatePickerModule,
                        ComboListModule,
                        FarrisSectionModule,
                        NumberSpinnerModule,
                        FarrisTooltipModule,
                        FarrisPanelModule,
                        GSPMetadataRTServiceModule.forRoot(""),
                    ],
                    providers: [
                        FrameworkService,
                        HttpService,
                        SessionService,
                        CacheService,
                        FrmEventBus,
                        FuncsService,
                        LoadingService,
                        NotifyService,
                        UbccParamDataManangerService,
                        UbccParamDataUiService,
                        UbccParamDataUtils,
                        BsModalService
                    ],
                    entryComponents: [
                        UbccParamDataUiDialogComponent,
                        UbccParamDataUiComponent,
                        UbccParamDimensionUiComponent,
                        UbccParamDimensionUiRootComponent,
                        UbccParamDataUiRootComponent
                    ],
                    exports: [
                        UbccParamDataUiRootComponent,
                        UbccParamDataUiComponent,
                        UbccParamDataUiDialogComponent,
                        UbccParamDimensionUiComponent,
                        UbccParamDimensionUiRootComponent
                    ]
                },] }
    ];
    return UbccParamDataUiRootModule;
}());
export { UbccParamDataUiRootModule };
