// @farris/extend-page-modal/esm2015/lib/page-modal-backdrop.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/page-modal-backdrop.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Renderer2 } from '@angular/core';
import { FEPageModalCLASS_NAME } from './page-modal-options.class';
import { Utils } from '@farris/ui-modal/utils';
/**
 * This component will be added as background layout for modals if enabled
 */
export class FEModalBackdropComponent {
    /**
     * @param {?} element
     * @param {?} renderer
     */
    constructor(element, renderer) {
        this._isShown = false;
        this._dialogType = 'modal';
        this.element = element;
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    get isAnimated() {
        return this._isAnimated;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set isAnimated(value) {
        this._isAnimated = value;
        // this.renderer.setElementClass(this.element.nativeElement, `${ClassName.FADE}`, value);
    }
    /**
     * @return {?}
     */
    get isShown() {
        return this._isShown;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set isShown(value) {
        this._isShown = value;
        if (value) {
            this.renderer.addClass(this.element.nativeElement, `${FEPageModalCLASS_NAME.IN}`);
        }
        else {
            this.renderer.removeClass(this.element.nativeElement, `${FEPageModalCLASS_NAME.IN}`);
        }
        if (value) {
            this.renderer.addClass(this.element.nativeElement, `${FEPageModalCLASS_NAME.SHOW}`);
        }
        else {
            this.renderer.removeClass(this.element.nativeElement, `${FEPageModalCLASS_NAME.SHOW}`);
        }
    }
    /**
     * @return {?}
     */
    get dialogType() {
        return this._dialogType;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set dialogType(value) {
        // 只有一次显示，不考虑其他
        if (value) {
            this.renderer.addClass(this.element.nativeElement, 'fe-modal-backdrop-' + value);
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.isAnimated) {
            this.renderer.addClass(this.element.nativeElement, `${FEPageModalCLASS_NAME.FADE}`);
            Utils.reflow(this.element.nativeElement);
        }
        this.isShown = true;
    }
}
FEModalBackdropComponent.decorators = [
    { type: Component, args: [{
                selector: 'fe-modal-backdrop',
                template: ' ',
                host: { class: FEPageModalCLASS_NAME.BACKDROP }
            }] }
];
/** @nocollapse */
FEModalBackdropComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
if (false) {
    /** @type {?} */
    FEModalBackdropComponent.prototype.element;
    /** @type {?} */
    FEModalBackdropComponent.prototype.renderer;
    /**
     * @type {?}
     * @protected
     */
    FEModalBackdropComponent.prototype._isAnimated;
    /**
     * @type {?}
     * @protected
     */
    FEModalBackdropComponent.prototype._isShown;
    /**
     * @type {?}
     * @protected
     */
    FEModalBackdropComponent.prototype._dialogType;
}
