// @farris/devkit/esm2015/lib/cache/provider/memory_cache_provider.js
import { CacheKeyCompare } from "../types";
export class MemoryCacheProvider {
    constructor(cacheKeyCompare) {
        this.store = new Array();
        this.cacheKeyCompare = cacheKeyCompare || new CacheKeyCompare();
    }
    has(key) {
        if (this.store.length < 1) {
            return false;
        }
        const index = this.store.findIndex((item) => item && this.cacheKeyCompare.compare(key, item.key));
        return index !== -1;
    }
    length() {
        return this.store.length;
    }
    set(value) {
        this.store.push(value);
    }
    get(key) {
        if (this.store.length < 1) {
            return undefined;
        }
        return this.store.find((item) => this.cacheKeyCompare.compare(key, item.key));
    }
    delete(key) {
        if (this.store.length < 1) {
            return;
        }
        const index = this.store.findIndex((item) => item && this.cacheKeyCompare.compare(key, item.key));
        this.store.splice(index, 1);
    }
    clear() {
        this.store.length = 0;
    }
    keys() {
        return this.store.keys();
    }
    values() {
        return this.store.values();
    }
}
