// @ecp-caf/caf-common/esm2015/lib/services/httpservice/http.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { SessionService } from '../sessionservice/session.service';
import { CommonUtil } from '../../utilities/common.utilities';
export class HttpService {
    /**
     * @param {?} http
     * @param {?} sessionService
     */
    constructor(http, sessionService) {
        this.http = http;
        this.sessionService = sessionService;
    }
    /**
     * @param {?} url
     * @return {?}
     */
    get(url) {
        return this.request('Get', url, this.setHeader({}));
    }
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    post(url, body) {
        return this.request('Post', url, this.setHeader({ body }));
    }
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    put(url, body) {
        return this.request('Put', url, this.setHeader({ body }));
    }
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    delete(url, body) {
        return this.request('Delete', url, this.setHeader({ body }));
    }
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    patch(url, body) {
        return this.request('Patch', url, this.setHeader({ body }));
    }
    /**
     * 默认用当前sessionid
     * @param {?} options
     * @param {?=} bizContextId
     * @return {?}
     */
    setHeader(options, bizContextId) {
        options.headers = {
            'Content-Type': 'application/json'
        };
        /** @type {?} */
        const commonVariables = CommonUtil.getCommonVariables();
        if (commonVariables) {
            /** @type {?} */
            const commonVariableKey = 'X-CAF-Runtime-CommonVariable';
            options.headers[commonVariableKey] = commonVariables;
        }
        if (bizContextId) {
            /** @type {?} */
            const bizContextKey = 'X-CAF-Runtime-Context';
            options.headers[bizContextKey] = bizContextId;
        }
        return options;
    }
    /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    request(method, url, options) {
        if (options.body) {
            if (typeof options.body !== 'string') {
                options.body = JSON.stringify(options.body);
            }
        }
        return this.http.request(method, url, options)
            .pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            return data;
        })));
    }
    /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    request$(method, url, options) {
        return this.http.request(method, url, options);
    }
}
HttpService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
HttpService.ctorParameters = () => [
    { type: HttpClient },
    { type: SessionService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    HttpService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    HttpService.prototype.sessionService;
}
