// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/constraint-serializer/decimal-constraint-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/constraint-serializer/decimal-constraint-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { UbccDecimalConstraint } from '../../extension/constraint/ubcc-decimal-constraint';
var UbccDecimalConstraintSerializer = /** @class */ (function () {
    function UbccDecimalConstraintSerializer() {
    }
    /**
     * @param {?} constraint
     * @param {?} ctxt
     * @return {?}
     */
    UbccDecimalConstraintSerializer.prototype.serialize = /**
     * @param {?} constraint
     * @param {?} ctxt
     * @return {?}
     */
    function (constraint, ctxt) {
        /** @type {?} */
        var jobj = new Object();
        if (!constraint) {
            return jobj;
        }
        /** @type {?} */
        var decimalConstraint = (/** @type {?} */ (constraint));
        jobj['kind'] = decimalConstraint.kind;
        jobj['max'] = decimalConstraint.max;
        jobj['min'] = decimalConstraint.min;
        jobj['precision'] = decimalConstraint.precision;
        jobj['unit'] = decimalConstraint.unit;
        return jobj;
    };
    /**
     * @param {?} jobj
     * @param {?} ctxt
     * @return {?}
     */
    UbccDecimalConstraintSerializer.prototype.deserialize = /**
     * @param {?} jobj
     * @param {?} ctxt
     * @return {?}
     */
    function (jobj, ctxt) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        var decimalConstraint = new UbccDecimalConstraint();
        decimalConstraint.max = (/** @type {?} */ (jobj['max']));
        decimalConstraint.min = (/** @type {?} */ (jobj['min']));
        decimalConstraint.precision = (/** @type {?} */ (jobj['precision']));
        decimalConstraint.unit = jobj['unit'];
        return decimalConstraint;
    };
    return UbccDecimalConstraintSerializer;
}());
export { UbccDecimalConstraintSerializer };
