// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/constraint-serializer/integer-constraint-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/constraint-serializer/integer-constraint-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { UbccIntegerConstraint } from '../../extension/constraint/ubcc-integer-constraint';
var UbccIntegerConstraintSerializer = /** @class */ (function () {
    function UbccIntegerConstraintSerializer() {
    }
    /**
     * @param {?} constraint
     * @param {?} ctxt
     * @return {?}
     */
    UbccIntegerConstraintSerializer.prototype.serialize = /**
     * @param {?} constraint
     * @param {?} ctxt
     * @return {?}
     */
    function (constraint, ctxt) {
        /** @type {?} */
        var jobj = new Object();
        if (!constraint) {
            return jobj;
        }
        /** @type {?} */
        var integerConstraint = (/** @type {?} */ (constraint));
        jobj['kind'] = integerConstraint.kind;
        jobj['max'] = integerConstraint.max;
        jobj['min'] = integerConstraint.min;
        jobj['unit'] = integerConstraint.unit;
        return jobj;
    };
    /**
     * @param {?} jobj
     * @param {?} ctxt
     * @return {?}
     */
    UbccIntegerConstraintSerializer.prototype.deserialize = /**
     * @param {?} jobj
     * @param {?} ctxt
     * @return {?}
     */
    function (jobj, ctxt) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        var integerConstraint = new UbccIntegerConstraint();
        integerConstraint.max = (/** @type {?} */ (jobj['max']));
        integerConstraint.min = (/** @type {?} */ (jobj['min']));
        integerConstraint.unit = jobj['unit'];
        return integerConstraint;
    };
    return UbccIntegerConstraintSerializer;
}());
export { UbccIntegerConstraintSerializer };
