// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/constraint-serializer/string-constraint-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/constraint-serializer/string-constraint-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { UbccStringConstraint } from '../../extension/constraint/ubcc-string-constaint';
var UbccStringConstraintSerializer = /** @class */ (function () {
    function UbccStringConstraintSerializer() {
    }
    /**
     * @param {?} constraint
     * @param {?} ctxt
     * @return {?}
     */
    UbccStringConstraintSerializer.prototype.serialize = /**
     * @param {?} constraint
     * @param {?} ctxt
     * @return {?}
     */
    function (constraint, ctxt) {
        /** @type {?} */
        var jobj = new Object();
        if (!constraint) {
            return jobj;
        }
        /** @type {?} */
        var stringConstraint = (/** @type {?} */ (constraint));
        jobj['kind'] = stringConstraint.kind;
        jobj['maxLength'] = stringConstraint.maxLength;
        jobj['minLength'] = stringConstraint.minLength;
        jobj['regexExpress'] = stringConstraint.regexExpress;
        return jobj;
    };
    /**
     * @param {?} jobj
     * @param {?} ctxt
     * @return {?}
     */
    UbccStringConstraintSerializer.prototype.deserialize = /**
     * @param {?} jobj
     * @param {?} ctxt
     * @return {?}
     */
    function (jobj, ctxt) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        var constaint = new UbccStringConstraint();
        constaint.maxLength = (/** @type {?} */ (jobj['maxLength']));
        constaint.minLength = (/** @type {?} */ (jobj['minLength']));
        constaint.regexExpress = jobj['regexExpress'];
        return constaint;
    };
    return UbccStringConstraintSerializer;
}());
export { UbccStringConstraintSerializer };
