// @gsp-wf/wf-process-editor/esm5/lib/views/assign-scope/other-assign-activity-node-info.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Condition } from "./condition";
var DisplayInfor = /** @class */ (function () {
    function DisplayInfor(name, id) {
        this.name = name;
        this.activityDefinitionId = id;
        this.required = false;
    }
    return DisplayInfor;
}());
export { DisplayInfor };
if (false) {
    /** @type {?} */
    DisplayInfor.prototype.name;
    /** @type {?} */
    DisplayInfor.prototype.activityDefinitionId;
    /** @type {?} */
    DisplayInfor.prototype.condition;
    /** @type {?} */
    DisplayInfor.prototype.required;
}
var OtherAssignActivityNodeInfo = /** @class */ (function () {
    function OtherAssignActivityNodeInfo(id, condition, required) {
        this.activityDefinitionId = id;
        this.condition = new Condition();
        this.condition.Value = condition;
        this.required = required;
    }
    return OtherAssignActivityNodeInfo;
}());
export { OtherAssignActivityNodeInfo };
if (false) {
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.activityDefinitionId;
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.condition;
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.required;
}
