// @farris/mobile-command-services/esm2015/lib/work-flow-service/wf-task-handler.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/work-flow-service/wf-task-handler.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
import { StartProcessPayload, CancelSubmitPayload } from './entity/start-process-payload';
/**
 * @record
 */
function Response() { }
if (false) {
    /** @type {?} */
    Response.prototype.type;
    /** @type {?} */
    Response.prototype.message;
}
;
export class WfTaskHandlerService {
    /**
     * @param {?} processSelectorService
     * @param {?} taskService
     * @param {?} loadingService
     * @param {?} notifyService
     */
    constructor(processSelectorService, taskService, loadingService, notifyService) {
        this.processSelectorService = processSelectorService;
        this.taskService = taskService;
        this.loadingService = loadingService;
        this.notifyService = notifyService;
        this.subject = new Subject();
    }
    /**
     * 提交审批
     * @param {?} payload 上下文参数
     * @return {?}
     */
    startProcess(payload) {
        /** @type {?} */
        const subject = new Subject();
        this.processSelectorService.setParam(payload);
        this.processSelectorService.getSubject().subscribe((/**
         * @param {?} response
         * @return {?}
         */
        (response) => {
            if (response.type === 'success') {
                if (response.message) {
                    this.notifyService.success(response.message);
                    subject.next(true);
                }
            }
            else if (response.type === 'error') {
                subject.next(false);
                this.notifyService.error(response.message);
            }
        }), (/**
         * @return {?}
         */
        () => {
            this.loadingService.hide();
        }));
        return subject.asObservable();
    }
    /**
     * 取消提交
     * \@payload payload 参数
     * @param {?} payload
     * @return {?}
     */
    cancelSubmit(payload) {
        /** @type {?} */
        const subject = new Subject();
        this.loadingService.show();
        this.taskService.cancelSubmit(payload).subscribe((/**
         * @param {?} re
         * @return {?}
         */
        re => {
            subject.next(true);
            this.loadingService.hide();
            this.notifyService.success('取消提交成功');
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            subject.next(false);
            this.loadingService.hide();
            this.notifyService.error(error && error.response && error.response.data && error.response.data.Message || '取消提交失败');
        }));
        return subject.asObservable();
    }
    /**
     * 提交审批
     * @param {?} dataId 表单id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    submitWithBizDefKey(dataId, bizDefKey) {
        if (!dataId) {
            this.notifyService.error('表单id不能为空');
            return;
        }
        if (!bizDefKey) {
            this.notifyService.error('入口单据Id不能为空');
            return;
        }
        /** @type {?} */
        var startProcessPayload = new StartProcessPayload();
        startProcessPayload.dataId = dataId;
        startProcessPayload.bizDefKey = bizDefKey;
        return this.startProcess(startProcessPayload);
    }
    /**
     * 取消提交审批
     * @param {?} dataId 表单id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    cancelSubmitWithDataId(dataId, bizDefKey) {
        if (!dataId) {
            this.notifyService.error('表单id不能为空');
            return;
        }
        if (!bizDefKey) {
            this.notifyService.error('入口单据Id不能为空');
            return;
        }
        /** @type {?} */
        var cancelSubmitPayload = new CancelSubmitPayload();
        cancelSubmitPayload.dataId = dataId;
        cancelSubmitPayload.bizDefKey = bizDefKey;
        return this.cancelSubmit(cancelSubmitPayload);
    }
    /**
     * 批量提交审批
     * @param {?} dataIds 表单ID集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    batchSubmitWithBizDefKey(dataIds, bizDefKey) {
        if (!dataIds) {
            this.notifyService.error('dataId不能为空');
            return;
        }
        if (!dataIds.length) {
            this.notifyService.error('请选择数据');
            return;
        }
        if (!bizDefKey) {
            this.notifyService.error('入口单据ID不能为空');
            return;
        }
        if (dataIds.length === 1) {
            /** @type {?} */
            var startPayload = new StartProcessPayload();
            startPayload.bizDefKey = bizDefKey;
            startPayload.dataId = dataIds[0];
            return this.startProcess(startPayload);
        }
        /** @type {?} */
        var successNum = 0;
        /** @type {?} */
        var failedNum = 0;
        /** @type {?} */
        const subject = new Subject();
        this.loadingService.show();
        for (var i = 0; i < dataIds.length; i++) {
            /** @type {?} */
            var startPayload = new StartProcessPayload();
            startPayload.bizDefKey = bizDefKey;
            startPayload.dataId = dataIds[i];
            startPayload.variables = { simulate: false };
            this.taskService.startProcess(startPayload).subscribe((/**
             * @param {?} response
             * @return {?}
             */
            (response) => {
                if (response.procDefs.length > 1 || response.nextFlowNodes.length > 1) { //存在多个流程，单独提交//多个节点，单独提交
                    failedNum++;
                }
                else {
                    successNum++;
                }
                if (successNum + failedNum === dataIds.length) {
                    this.loadingService.hide();
                    subject.next(true); //全部执行完再返回
                    this.showBatchSubmitInfo("submit", successNum, failedNum);
                }
            }), (/**
             * @return {?}
             */
            () => {
                failedNum++;
                if (successNum + failedNum === dataIds.length) {
                    this.loadingService.hide();
                    subject.next(false); //全部执行完再返回
                    this.showBatchSubmitInfo("submit", successNum, failedNum);
                }
            }));
        }
        return subject.asObservable();
    }
    /**
     * 批量取消提交审批
     * @param {?} dataIds 表单ID集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    batchCancelSubmitWithDataId(dataIds, bizDefKey) {
        if (!dataIds) {
            this.notifyService.error('dataId不能为空');
            return;
        }
        if (!dataIds.length) {
            this.notifyService.error('请选择数据');
            return;
        }
        if (!bizDefKey) {
            this.notifyService.error('入口单据ID不能为空');
            return;
        }
        /** @type {?} */
        const subject = new Subject();
        /** @type {?} */
        var successNum = 0;
        /** @type {?} */
        var failedNum = 0;
        this.loadingService.show();
        for (var i = 0; i < dataIds.length; i++) {
            /** @type {?} */
            var cancelSubmitPayload = new CancelSubmitPayload();
            cancelSubmitPayload.dataId = dataIds[i];
            cancelSubmitPayload.bizDefKey = bizDefKey;
            this.taskService.cancelSubmit(cancelSubmitPayload).subscribe((/**
             * @param {?} re
             * @return {?}
             */
            re => {
                successNum++;
                if (successNum + failedNum === dataIds.length) {
                    this.loadingService.hide();
                    subject.next(true);
                    this.showBatchSubmitInfo("cancel", successNum, failedNum);
                }
            }), (/**
             * @return {?}
             */
            () => {
                failedNum++;
                if (successNum + failedNum === dataIds.length) {
                    this.loadingService.hide();
                    subject.next(false);
                    this.showBatchSubmitInfo("cancel", successNum, failedNum);
                }
            }));
        }
        return subject.asObservable();
    }
    /**
     * 子表提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataId 子表id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    childSubmit(parentDataId, childDataId, bizDefKey) {
        if (!parentDataId || !childDataId) {
            this.notifyService.error('dataId不能为空');
            return;
        }
        if (!bizDefKey) {
            this.notifyService.error('入口单据ID不能为空');
            return;
        }
        /** @type {?} */
        var startProcessPayload = new StartProcessPayload();
        startProcessPayload.dataId = `${parentDataId},${childDataId}`;
        startProcessPayload.bizDefKey = bizDefKey;
        return this.startProcess(startProcessPayload);
    }
    /**
     * 子表取消提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataId 子表id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    childCancelSubmit(parentDataId, childDataId, bizDefKey) {
        if (!parentDataId || !childDataId) {
            this.notifyService.error('dataId不能为空');
            return;
        }
        if (!bizDefKey) {
            this.notifyService.error('入口单据ID不能为空');
            return;
        }
        /** @type {?} */
        var cancelSubmitPayload = new CancelSubmitPayload();
        cancelSubmitPayload.dataId = `${parentDataId},${childDataId}`;
        cancelSubmitPayload.bizDefKey = bizDefKey;
        return this.cancelSubmit(cancelSubmitPayload);
    }
    /**
     * 子表批量提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataIds 子表id集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    childBatchSubmit(parentDataId, childDataIds, bizDefKey) {
        if (!parentDataId || !childDataIds) {
            this.notifyService.error('dataId不能为空');
            return;
        }
        if (!bizDefKey) {
            this.notifyService.error('入口单据ID不能为空');
            return;
        }
        /** @type {?} */
        const dataIds = childDataIds.map((/**
         * @param {?} childDataId
         * @return {?}
         */
        childDataId => `${parentDataId},${childDataId}`));
        return this.batchSubmitWithBizDefKey(dataIds, bizDefKey);
    }
    /**
     * 子表批量取消提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataIds 子表id集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    childBatchCancelSubmit(parentDataId, childDataIds, bizDefKey) {
        if (!parentDataId || !childDataIds) {
            this.notifyService.error('dataId不能为空');
            return;
        }
        if (!bizDefKey) {
            this.notifyService.error('入口单据ID不能为空');
            return;
        }
        /** @type {?} */
        const dataIds = childDataIds.map((/**
         * @param {?} childDataId
         * @return {?}
         */
        childDataId => `${parentDataId},${childDataId}`));
        return this.batchCancelSubmitWithDataId(dataIds, bizDefKey);
    }
    /**
     * 批量提交、取消提交提示信息
     * @private
     * @param {?} actionCode
     * @param {?} successNum
     * @param {?} failedNum
     * @return {?}
     */
    showBatchSubmitInfo(actionCode, successNum, failedNum) {
        if (successNum && !failedNum) { //批量全部成功
            actionCode === "submit" && this.notifyService.success('提交审批成功！');
            actionCode === "cancel" && this.notifyService.success('取消提交成功！');
        }
        if (successNum && failedNum) { //批量有成功有失败
            this.notifyService.warning(`操作成功${successNum}条，失败${failedNum}条!`);
        }
        if (!successNum && failedNum) { //批量全部失败
            actionCode === "submit" && this.notifyService.error('提交审批失败！');
            actionCode === "cancel" && this.notifyService.error('取消提交失败！');
        }
    }
}
if (false) {
    /** @type {?} */
    WfTaskHandlerService.prototype.subject;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.processSelectorService;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.taskService;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.loadingService;
    /**
     * @type {?}
     * @private
     */
    WfTaskHandlerService.prototype.notifyService;
}
