// @gspwidget/widget-core/esm5/lib/widget/runtime-agg.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PlThemeColors } from "@gspwidget/util";
/** @type {?} */
var aggFirstDim2Field = (/**
 * @param {?} runtimeAgg
 * @param {?} allDims
 * @return {?}
 */
function (runtimeAgg, allDims) { return getFieldByName(allDims, runtimeAgg.dimensions[0]); });
var ɵ0 = aggFirstDim2Field;
/** @type {?} */
var aggFirstInd2Field = (/**
 * @param {?} runtimeAgg
 * @param {?} allDims
 * @return {?}
 */
function (runtimeAgg, allDims) { return getFieldByName(allDims, runtimeAgg.indicators[0].dimName); });
var ɵ1 = aggFirstInd2Field;
/** @type {?} */
var stackAggFields = (/**
 * @param {?=} pure
 * @return {?}
 */
function (pure) {
    if (pure === void 0) { pure = false; }
    return [{
            name: 'yFieldObjs',
            valueFactory: (/**
             * @param {?} __0
             * @param {?} allDims
             * @return {?}
             */
            function (_a, allDims) {
                var indicators = _a.indicators;
                return indicators.map((/**
                 * @param {?} ind
                 * @param {?} i
                 * @return {?}
                 */
                function (ind, i) { return ({
                    valueField: getFieldByName(allDims, ind.dimName),
                    color: pure
                        ? PlThemeColors.pure[i % indicators.length]
                        : PlThemeColors.gradient[i % indicators.length]
                }); }));
            })
        }, {
            name: 'categoryFieldName',
            valueFactory: aggFirstDim2Field
        }];
});
var ɵ2 = stackAggFields;
/** @type {?} */
var barAggFields = [
    { name: 'categoryFieldName', valueFactory: aggFirstDim2Field },
    { name: 'valueFieldName', valueFactory: aggFirstInd2Field }
];
/** @type {?} */
export var chartAggFieldMap = {
    stack: stackAggFields(),
    area: stackAggFields(true),
    'bar-group': stackAggFields(),
    'widget-line': stackAggFields(true),
    bar: barAggFields,
    pie: barAggFields,
    pie2: barAggFields,
    'bar-horizontal': barAggFields,
    'simple-line': barAggFields,
    'bar-multiple': [{
            categoryFieldName: aggFirstDim2Field,
            valueFieldNameOne: aggFirstInd2Field,
            valueFieldNameTwo: (/**
             * @param {?} runtimeAgg
             * @param {?} allDims
             * @return {?}
             */
            function (runtimeAgg, allDims) { return getFieldByName(allDims, runtimeAgg.indicators[1].dimName); }),
        }]
};
/**
 * @param {?} dims
 * @param {?} dimName
 * @return {?}
 */
function getFieldByName(dims, dimName) {
    var _a = dims.find((/**
     * @param {?} dim
     * @return {?}
     */
    function (dim) { return dim.name === dimName; })), name = _a.name, label = _a.label;
    return { name: name, label: label };
}
/**
 * @param {?} widgetName
 * @return {?}
 */
export function isWidgetRuntimeAggActive(widgetName) {
    return Object
        .keys(chartAggFieldMap)
        .includes(widgetName);
}
export { ɵ0, ɵ1, ɵ2 };
