// @gsp-cmp/common-component/esm2015/serialize-context/default/default-gsp-component-serialize-context.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DefaultOperationSerializerFactory, DefaultParameterSerializerFactory, DefaultPropertySerializerFactory } from '@ecp-caf/common-structure';
import { DefaultGspComponentDataTypeSerializerFactory } from './factory/default-gsp-component-data-type-serializer-factory';
export class DefaultGspComponentSerializeContext {
    // private gspComponentFactory: IGspComponentSerializerFactory;
    /**
     * @return {?}
     */
    getDataTypeSerializerFactory() {
        if (!this.dataTypeFactory) {
            this.dataTypeFactory = new DefaultGspComponentDataTypeSerializerFactory();
        }
        return this.dataTypeFactory;
    }
    /**
     * @return {?}
     */
    getOperationSerializerFactory() {
        if (!this.operationFactory) {
            this.operationFactory = new DefaultOperationSerializerFactory();
        }
        return this.operationFactory;
    }
    /**
     * @return {?}
     */
    getParameterSerializerFactory() {
        if (!this.parameterFactory) {
            this.parameterFactory = new DefaultParameterSerializerFactory();
        }
        return this.parameterFactory;
    }
    /**
     * @return {?}
     */
    getPropertySerializerFactory() {
        if (!this.propertyFactory) {
            this.propertyFactory = new DefaultPropertySerializerFactory();
        }
        return this.propertyFactory;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    DefaultGspComponentSerializeContext.prototype.dataTypeFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultGspComponentSerializeContext.prototype.operationFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultGspComponentSerializeContext.prototype.parameterFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultGspComponentSerializeContext.prototype.propertyFactory;
}
