// @farris/devkit/esm2015/lib/rest/rest_service.js
/* eslint-disable new-cap */
/* eslint-disable valid-jsdoc */
/* eslint-disable require-jsdoc */
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { BasePathService } from '@farris/rtf';
export class RestfulService {
    /**
     * 构造函数
     * @param httpClient http客户端
     */
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * 发送GET请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    get(uri, params, options) {
        return this.request(uri, 'GET', params, options);
    }
    /**
     * 发送POST请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    put(uri, body, params, options) {
        const optionsWithBody = this.addBody(options, body);
        return this.request(uri, 'PUT', params, optionsWithBody);
    }
    /**
     * 发送PUT请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    post(uri, body, params, options) {
        const optionsWithBody = this.addBody(options, body);
        return this.request(uri, 'POST', params, optionsWithBody);
    }
    /**
     * 发送DELETE请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    delete(uri, params, options) {
        return this.request(uri, 'DELETE', params, options);
    }
    /**
     * 发送请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    request(uri, method, params, options = {}) {
        options = options || {};
        if (params) {
            const httpParams = this.buildParams(params);
            options.params = httpParams;
        }
        const methodName = method;
        uri = BasePathService.convertPath(uri);
        const result = this.httpClient.request(methodName, uri, options);
        return result;
    }
    /**
     * 构造Http请求参数
     * @param params 请求参数
     */
    buildParams(params) {
        let httpParams = new HttpParams();
        for (const key in params) {
            if (params.hasOwnProperty(key)) {
                const value = params[key].toString();
                httpParams = httpParams.append(key, value);
            }
        }
        return httpParams;
    }
    addBody(options, body) {
        options = options || {};
        const mergedOptions = Object.assign(options, { body: body });
        return mergedOptions;
    }
}
RestfulService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RestfulService.ctorParameters = () => [
    { type: HttpClient }
];
