// @gsp-cmp/ubcc-param-vo/esm2015/lib/vo/serializer/dimension-serializer/dimension-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/dimension-serializer/dimension-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Dimension } from '../../param-dimension/dimension';
import { DimensionItemSerializer } from './dimension-item-serializer';
export class DimensionSerializer {
    /**
     * @param {?} dimension
     * @return {?}
     */
    serialize(dimension) {
        /** @type {?} */
        let jobj = new Object();
        if (!dimension) {
            return null;
        }
        jobj['id'] = dimension.id;
        jobj['code'] = dimension.code;
        jobj['name'] = dimension.name;
        if (dimension.dimensionItems) {
            jobj['dimensionItems'] = [];
            /** @type {?} */
            let serializer = new DimensionItemSerializer();
            dimension.dimensionItems.forEach((/**
             * @param {?} dimensionItem
             * @return {?}
             */
            dimensionItem => {
                jobj['dimensionItems'].push(serializer.serialize(dimensionItem));
            }));
        }
        return jobj;
    }
    /**
     * @param {?} jobj
     * @return {?}
     */
    deserialize(jobj) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        let dimension = new Dimension();
        dimension.id = jobj['id'];
        dimension.code = jobj['code'];
        dimension.name = jobj['name'];
        if (jobj['dimensionItems']) {
            dimension.dimensionItems = new Array();
            /** @type {?} */
            let serializer = new DimensionItemSerializer();
            jobj['dimensionItems'].forEach((/**
             * @param {?} dimensionItemJobj
             * @return {?}
             */
            dimensionItemJobj => {
                dimension.dimensionItems.push(serializer.deserialize(dimensionItemJobj));
            }));
        }
        return dimension;
    }
}
