// @farris/ui-wizard/esm2015/lib/page/page.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, Output, EventEmitter } from '@angular/core';
import { WizardComponent } from '../wizard/wizard.component';
import { PageService } from '../service/page.service';
export class FarrisWizardPageComponent {
    /**
     * @param {?} wizardSet
     * @param {?} pageSer
     */
    constructor(wizardSet, pageSer) {
        this.pageSer = pageSer;
        // 唯一标记
        this.pageId = '';
        // 默认class
        this.addClass = true;
        this._active = false;
        this.pageSelected = new EventEmitter();
        this.wizardSet = wizardSet;
        this.wizardSet.addPage(this);
    }
    /**
     * @return {?}
     */
    get activeCls() {
        return this._active;
    }
    // 当前选中
    /**
     * @param {?} value
     * @return {?}
     */
    set active(value) {
        this._active = value;
        if (this.pageId) {
            this.doAfterActive();
        }
    }
    /**
     * @return {?}
     */
    get active() {
        return this._active;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        //console.log('FarrisWizardPageComponent——' + this.pageId);
        this.pageSer.setPageId(this.pageId);
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.doAfterActive();
    }
    /**
     * @return {?}
     */
    getPageId() {
        return this.pageId;
    }
    /**
     *
     * @param {?} value
     * @return {?}
     */
    setPageActive(value) {
        if (this.active !== value) {
            this.active = value;
        }
    }
    /**
     * 处理Active
     * @return {?}
     */
    doAfterActive() {
        this.pageSelected.emit({
            pageId: this.pageId,
            state: this.active
        });
        //console.log('FarrisWizardPageComponent——' + this.pageid + this.active);
        this.pageSer.activeState.next({
            pageId: this.pageId,
            state: this.active
        });
    }
}
FarrisWizardPageComponent.decorators = [
    { type: Component, args: [{
                selector: 'wizard-page',
                template: '<ng-content></ng-content>',
                providers: [PageService],
                styles: [":host.f-component-wizard-page{display:none;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;-webkit-box-flex:1;flex-grow:1;flex-shrink:1;flex-basis:0;position:relative}:host.f-component-wizard-page.active{display:-webkit-box;display:flex}"]
            }] }
];
/** @nocollapse */
FarrisWizardPageComponent.ctorParameters = () => [
    { type: WizardComponent },
    { type: PageService }
];
FarrisWizardPageComponent.propDecorators = {
    pageId: [{ type: Input }],
    addClass: [{ type: HostBinding, args: ['class.f-component-wizard-page',] }],
    activeCls: [{ type: HostBinding, args: ['class.active',] }],
    active: [{ type: Input }],
    pageSelected: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FarrisWizardPageComponent.prototype.pageId;
    /** @type {?} */
    FarrisWizardPageComponent.prototype.addClass;
    /**
     * @type {?}
     * @private
     */
    FarrisWizardPageComponent.prototype._active;
    /** @type {?} */
    FarrisWizardPageComponent.prototype.pageSelected;
    /** @type {?} */
    FarrisWizardPageComponent.prototype.wizardSet;
    /**
     * @type {?}
     * @private
     */
    FarrisWizardPageComponent.prototype.pageSer;
}
