// @farris/ui-personnel-selector/esm5/utils/locale.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/locale.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var PERSONNEL_SELECTOR_LOCALE = {
    "zh-CHS": {
        personnelSelector: {
            selection: {
                selectionDialogTitle: '人员选择',
                recent: '最近',
                candidate: '人员',
                organization: '组织架构',
                sysUserGroup: '用户分组',
                group: '我的群组',
                favorites: '我的收藏',
                searchPlaceHolder: '请输入姓名汉字/首字母/编号搜索',
                confirm: '确定',
                cancel: '取消',
                select: '请从左侧选择人员数据'
            },
            tabCommon: {
                checkAll: '全选',
                noUserData: '暂无数据',
                clearAll: '清空联系人'
            },
            organization: {
                searchPlaceHolder: '输入关键字定位',
                noDataInfo: '未找到相关组织',
                unSelectableInfo: '相关组织不在可定位的范围内',
                enableChildHierarchy: '含子级',
                includeCurrentUser: '含本人',
                childHierarchyTitle: '含子级',
                currentUserTitle: '含本人'
            },
            userGroup: {
                searchPlaceHolder: '输入关键字定位'
            },
            group: {
                newGroupName: '群组名称',
                newGroupNote: '群组备注',
                newGroupPlaceHolder: '请输入',
                publicGroup: '公共群组',
                privateGroup: '个人群组',
                appendMembers: '添加组员',
                items: '人',
                saveGroup: '保存为群组',
                editGroup: '编辑群组',
                noPersonnelMessage: '请先选择人员。'
            },
            filter: {
                title: '筛选',
                showLeader: '显示集团领导',
                position: '岗位',
                organization: '组织',
                selectPlaceHolder: '请选择',
                noOrgData: '暂无数据',
                clearAll: '清空已选'
            },
            searchPage: {
                firstResultText: '搜索结果 (',
                lastResultText: ')',
                close: '关闭'
            },
            quickSelect: {
                recent: '最近',
                result: '搜索结果',
                firstTotalText: '已选',
                lastTotalText: '人',
                quickMode: '快捷模式',
                advancedMode: '高级模式'
            },
            message: {
                mapFields: '字段映射中未添加用户唯一标识。',
                groupExtend: '群组扩展函数不存在。',
                inappropriateData: '数据问题',
                vaildElements: '前2个参数必填',
                multiSelect: '单选模式下预置了多条选中数据，请修改！',
                filterOrgs: '筛选组织：',
                noneFilterOrgs: '筛选组织：无',
                filterPositionError: '根据组织和岗位获取人员数据时，调用后端接口出现错误！',
                filterOrgError: '显示组织过滤漏斗的场景下，调用后端接口出现错误！'
            }
        }
    },
    "zh-CHT": {
        personnelSelector: {
            selection: {
                selectionDialogTitle: '人員選擇',
                recent: '最近',
                candidate: '人員',
                organization: '組織架構',
                sysUserGroup: '用戶分組',
                group: '我的群組',
                favorites: '我的收藏',
                searchPlaceHolder: '請輸入姓名漢字/首字母/編號搜索',
                confirm: '確定',
                cancel: '取消',
                select: '請從左側選擇人員數據'
            },
            tabCommon: {
                checkAll: '全選',
                noUserData: '暫無數據',
                clearAll: '清空聯係人'
            },
            organization: {
                searchPlaceHolder: '輸入關鍵字定位',
                noDataInfo: '未找到相關組織',
                unSelectableInfo: '相關組織不在可定位的範圍内',
                enableChildHierarchy: '含子級',
                includeCurrentUser: '含本人',
                childHierarchyTitle: '含子級',
                currentUserTitle: '含本人'
            },
            userGroup: {
                searchPlaceHolder: '輸入關鍵字定位'
            },
            group: {
                newGroupName: '群組名稱',
                newGroupNote: '群組備注',
                newGroupPlaceHolder: '請輸入',
                publicGroup: '公共群組',
                privateGroup: '個人群組',
                appendMembers: '添加組員',
                items: '人',
                saveGroup: '保存爲群組',
                editGroup: '編輯群組',
                noPersonnelMessage: '請先選擇人員。'
            },
            filter: {
                title: '篩選',
                showLeader: '顯示集團領導',
                position: '崗位',
                organization: '組織',
                selectPlaceHolder: '請選擇',
                noOrgData: '暫無數據',
                clearAll: '清空已選'
            },
            searchPage: {
                firstResultText: '搜索結果 (',
                lastResultText: ')',
                close: '關閉'
            },
            quickSelect: {
                recent: '最近',
                result: '搜索結果',
                firstTotalText: '已選',
                lastTotalText: '人',
                quickMode: '快捷模式',
                advancedMode: '高級模式'
            },
            message: {
                mapFields: '字段映射中未添加用戶唯一標識。',
                groupExtend: '群組擴展函數不存在。',
                inappropriateData: '數據問題',
                vaildElements: '前2個參數必填',
                multiSelect: '單選模式下預置了多條選中數據，請修改！',
                filterOrgs: '篩選組織：',
                noneFilterOrgs: '篩選組織：無',
                filterPositionError: '根據組織和崗位獲取人員數據時，調用後端接口出現錯誤！',
                filterOrgError: '顯示組織過濾漏斗的場景下，調用後端接口出現錯誤！'
            }
        }
    },
    en: {
        personnelSelector: {
            selection: {
                selectionDialogTitle: 'Select Personnel',
                recent: 'Recent',
                candidate: 'Candidates',
                organization: 'Organizations',
                sysUserGroup: 'User Groups',
                group: 'Groups',
                favorites: 'Favorites',
                searchPlaceHolder: 'Search Personnel By Name, First Phonetic Letter or Code',
                confirm: 'Confirm',
                cancel: 'Cancel',
                select: 'Please Select Personnel'
            },
            tabCommon: {
                checkAll: 'All',
                noUserData: 'Empty Data',
                clearAll: 'Clear'
            },
            organization: {
                searchPlaceHolder: 'Search Organizations',
                noDataInfo: 'Empty Related Organization',
                unSelectableInfo: 'Related Organization Cannot be Selected',
                enableChildHierarchy: 'Subordinate',
                includeCurrentUser: 'User',
                childHierarchyTitle: 'Display Members of Subordinate Hierarchy',
                currentUserTitle: 'Display Current User'
            },
            userGroup: {
                searchPlaceHolder: 'Search UserGroups'
            },
            group: {
                newGroupName: 'Name',
                newGroupNote: 'Note',
                newGroupPlaceHolder: 'Please input',
                publicGroup: 'Pulic Group',
                privateGroup: 'Private Group',
                appendMembers: 'Append Members',
                items: 'items',
                saveGroup: 'Save Group',
                editGroup: 'Edit Group',
                noPersonnelMessage: 'Please select personnels firstly.'
            },
            filter: {
                title: 'Filter',
                showLeader: 'Show Leader',
                position: 'Position',
                organization: 'Organization',
                selectPlaceHolder: 'Please select',
                noOrgData: 'Empty Data',
                clearAll: 'Clear'
            },
            searchPage: {
                firstResultText: 'Search Result (',
                lastResultText: ')',
                close: 'Close'
            },
            quickSelect: {
                recent: 'Recent',
                result: 'Search Result',
                firstTotalText: 'Total',
                lastTotalText: 'items',
                quickMode: 'Quick',
                advancedMode: 'Advanced'
            },
            message: {
                mapFields: 'UserId is not included in mapping fields.',
                groupExtend: 'Extend function of group is not existed.',
                inappropriateData: 'Error Data',
                vaildElements: 'The first two elements are required',
                multiSelect: 'More than one item are selected in single-select mode, please correct!',
                filterOrgs: 'Filter Organization: ',
                noneFilterOrgs: 'Filter Organization: None',
                filterPositionError: 'Failed to get positions in filter personnels secondly!',
                filterOrgError: 'Failed to get organizations in filter personnels secondly!'
            }
        }
    }
};
