// @farris/ui-responsive/esm2015/lib/row/res-row.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, HostBinding } from '@angular/core';
export class FarrisRowDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        this.type = 'flex';
        this.gutter = 16;
        this.justify = 'start';
        this.align = 'start';
        // 绑定row 到class
        this.isRow = this.type === 'flex';
        //
        this.nativeClass = '';
        // 初始化类名 添加宿主元素已有的类名
        this.nativeClass += this.el.nativeElement.classList.value;
    }
    /**
     * @return {?}
     */
    get respClass() {
        return `${this.getJustifyClass()}${this.getAlignClass()}${this.getGutterClass()}${this.getBreakpointClass()}`;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    // 获取justify class
    /**
     * @return {?}
     */
    getJustifyClass() {
        /** @type {?} */
        const justifyStr = this.justify !== 'start' ? ` justify-content-${this.justify}` : ' ';
        return `${this.nativeClass}${justifyStr}`;
    }
    // 获取 align class
    /**
     * @return {?}
     */
    getAlignClass() {
        return this.align !== 'start' ? ` align-items-${this.align}` : ' ';
    }
    // 响应式 justify
    /**
     * @return {?}
     */
    getBreakpointClass() {
        /** @type {?} */
        const makeClass = (/**
         * @param {?} breakpoint
         * @return {?}
         */
        (breakpoint) => {
            /** @type {?} */
            const props = this[breakpoint] || {};
            // 获取对象中  justify或者align
            return Object.keys(props).map((/**
             * @param {?} prop
             * @return {?}
             */
            prop => prop === 'justify' ?
                `justify-content-${breakpoint}-${props[prop]}` : `align-items-${breakpoint}-${props[prop]}`)).join(' ');
        });
        /** @type {?} */
        const classStr = ['sm', 'md', 'lg', 'xl'].reduce((/**
         * @param {?} pre
         * @param {?} current
         * @return {?}
         */
        (pre, current) => {
            return `${pre} ${makeClass(current)}`;
        }), '');
        return classStr;
    }
    // 有无间隔
    /**
     * @return {?}
     */
    getGutterClass() {
        return !this.gutter ? ' no-gutters' : ' ';
    }
}
FarrisRowDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-row]',
            },] }
];
/** @nocollapse */
FarrisRowDirective.ctorParameters = () => [
    { type: ElementRef }
];
FarrisRowDirective.propDecorators = {
    type: [{ type: Input }],
    gutter: [{ type: Input }],
    justify: [{ type: Input }],
    align: [{ type: Input }],
    xs: [{ type: Input }],
    sm: [{ type: Input }],
    md: [{ type: Input }],
    lg: [{ type: Input }],
    xl: [{ type: Input }],
    respClass: [{ type: HostBinding, args: ['class',] }],
    isRow: [{ type: HostBinding, args: ['class.row',] }]
};
if (false) {
    /** @type {?} */
    FarrisRowDirective.prototype.type;
    /** @type {?} */
    FarrisRowDirective.prototype.gutter;
    /** @type {?} */
    FarrisRowDirective.prototype.justify;
    /** @type {?} */
    FarrisRowDirective.prototype.align;
    /**
     * xs={
     * justify:'start',
     * align:'start'
     * }
     * @type {?}
     */
    FarrisRowDirective.prototype.xs;
    /** @type {?} */
    FarrisRowDirective.prototype.sm;
    /** @type {?} */
    FarrisRowDirective.prototype.md;
    /** @type {?} */
    FarrisRowDirective.prototype.lg;
    /** @type {?} */
    FarrisRowDirective.prototype.xl;
    /** @type {?} */
    FarrisRowDirective.prototype.isRow;
    /** @type {?} */
    FarrisRowDirective.prototype.nativeClass;
    /**
     * @type {?}
     * @private
     */
    FarrisRowDirective.prototype.el;
}
