// @farris/mobile-command-services/esm2015/lib/attachment-services/attachment.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/attachment-services/attachment.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EMPTY, forkJoin } from 'rxjs';
import { tap } from 'rxjs/operators';
import { BindingPathConverter } from '@farris/mobile-devkit';
import { AttachmentUtil } from './attachment.util';
/**
 * 附件服务
 */
class AttachmentService {
    /**
     * 构造函数
     * @param {?} viewModelContext
     * @param {?} entityService
     * @param {?} removeDataService
     * @param {?} attachDataService
     */
    constructor(viewModelContext, entityService, removeDataService, attachDataService) {
        /**
         * 默认根目录
         */
        this.defaultRootDirId = 'default-root';
        this.viewModelContext = viewModelContext;
        this.attachDataService = attachDataService;
        this.entityService = entityService;
        this.removeDataService = removeDataService;
    }
    /**
     * 默认子目录
     * @private
     * @return {?}
     */
    get defaultParentDirName() {
        return this.viewModelContext.bindingData.list.currentId;
    }
    /**
     * 上传单个文件
     * @param {?} attachmentInfoFieldPath
     * @param {?=} rootDirId
     * @param {?=} parentDirName
     * @return {?}
     */
    uploadAndUpdateRow(attachmentInfoFieldPath, rootDirId, parentDirName) {
        /** @type {?} */
        const rootId = rootDirId ? rootDirId : this.defaultRootDirId;
        /** @type {?} */
        const formId = parentDirName ? parentDirName : this.defaultParentDirName;
        if (!rootId || !formId) {
            throw new Error('rootDirId和parentDirName不能为空，请填写');
        }
        /** @type {?} */
        const fileInfos = this.getUploadFileInfosFromContext();
        /** @type {?} */
        const attachmentInfos = AttachmentUtil.convertToAttachmentInfos(fileInfos);
        if (!attachmentInfos || attachmentInfos.length === 0) {
            alert('请先上传附件');
            return EMPTY;
        }
        // 更新服务器端
        console.log('show loading ...');
        /** @type {?} */
        const firstAttachmentInfo = AttachmentUtil.getFirstAttachmentInfo(attachmentInfos);
        /** @type {?} */
        const result$ = this.attachDataService.updateRow(attachmentInfoFieldPath, firstAttachmentInfo).pipe(tap((/**
         * @return {?}
         */
        () => {
            console.log('hide loading ...');
        })));
        return result$;
    }
    /**
     * 上传多个文件
     * @param {?} attachmentInfoFieldPath
     * @param {?=} rootDirId
     * @param {?=} parentDirName
     * @return {?}
     */
    uploadAndBatchAddRows(attachmentInfoFieldPath, rootDirId, parentDirName) {
        /** @type {?} */
        const rootId = rootDirId ? rootDirId : this.defaultRootDirId;
        /** @type {?} */
        const formId = parentDirName ? parentDirName : this.defaultParentDirName;
        if (!rootId || !formId) {
            throw new Error('rootDirId和parentDirName不能为空，请填写');
        }
        /** @type {?} */
        const fileInfos = this.getUploadFileInfosFromContext();
        /** @type {?} */
        const attachmentInfos = AttachmentUtil.convertToAttachmentInfos(fileInfos);
        if (!attachmentInfos || attachmentInfos.length === 0) {
            alert('请先上传附件');
            return EMPTY;
        }
        // 更新服务器端
        console.log('show loading ...');
        /** @type {?} */
        const result$ = this.attachDataService.updateRows(attachmentInfoFieldPath, attachmentInfos).pipe(tap((/**
         * @return {?}
         */
        () => {
            console.log('hide loading ...');
        })));
        return result$;
    }
    /**
     * 批量删除附件所在的行
     * @param {?} attachmentInfoFieldPath
     * @return {?}
     */
    removeAttachmentRows(attachmentInfoFieldPath) {
        if (!attachmentInfoFieldPath) {
            return EMPTY;
        }
        /** @type {?} */
        const dataIds = this.getDataIdsToRemove(attachmentInfoFieldPath);
        //如果是主表
        if (attachmentInfoFieldPath.split('/').length <= 2) {
            return this.removeDataService.removeByIds(dataIds);
        }
        // 如果是子表
        else {
            /** @type {?} */
            const removeObservables = [];
            if (dataIds.length === 0) {
                alert('请选择要删除的文件');
            }
            /** @type {?} */
            const bindingListPath = this.getBindingListPathWithAttachments(attachmentInfoFieldPath);
            dataIds.forEach((/**
             * @param {?} dataId
             * @return {?}
             */
            (dataId) => {
                /** @type {?} */
                const removeObservable = this.removeDataService.removeByPathAndId(bindingListPath, dataId);
                removeObservables.push(removeObservable);
            }));
            return forkJoin(removeObservables);
        }
    }
    // #region 工具方法
    /**
     * 从上下文中中获取控件传递的附件信息
     * @private
     * @return {?}
     */
    getUploadFileInfosFromContext() {
        /** @type {?} */
        const commandContext = (/** @type {?} */ (this['context']));
        /** @type {?} */
        const uploadFileInfos = (/** @type {?} */ (commandContext.eventParams));
        if (!uploadFileInfos) {
            return [];
        }
        return uploadFileInfos;
    }
    /**
     * 获取要删除的附件对应的数据id数组
     * @param {?} attachmentInfoFieldPath
     * @return {?}
     */
    getDataIdsToRemove(attachmentInfoFieldPath) {
        /** @type {?} */
        const attachIds = this.getAttachmentIdsToRemoveFromContext();
        /** @type {?} */
        const dataIds = [];
        attachIds.forEach((/**
         * @param {?} attachId
         * @return {?}
         */
        (attachId) => {
            // 上传删除和预览删除传递过来的fileId的key可能不一致，要做兼容
            /** @type {?} */
            const dataId = this.convertAttachmentIdToDataId(attachId, attachmentInfoFieldPath);
            dataIds.push(dataId);
        }));
        return dataIds;
    }
    /**
     * 从命令上下文中获取要删除附件ids
     * @return {?}
     */
    getAttachmentIdsToRemoveFromContext() {
        /** @type {?} */
        const commandContext = (/** @type {?} */ (this['context']));
        return (/** @type {?} */ (commandContext.eventParams));
    }
    /**
     * 根据路径获取附件字段值数组
     * @private
     * @param {?} fileId
     * @param {?} attachmentInfoFieldPath
     * @return {?}
     */
    convertAttachmentIdToDataId(fileId, attachmentInfoFieldPath) {
        // 解析路径
        /** @type {?} */
        const attachInfoBindingPath = BindingPathConverter.toBindingPathArray(attachmentInfoFieldPath);
        /** @type {?} */
        const attachFieldName = attachInfoBindingPath.pop();
        /** @type {?} */
        const attachListBindingPath = attachInfoBindingPath;
        // 获取附件id数组
        /** @type {?} */
        const entityListData = this.entityService.getEntityListData(attachListBindingPath);
        /** @type {?} */
        const targetEntityData = entityListData.find((/**
         * @param {?} entityData
         * @return {?}
         */
        (entityData) => {
            if (entityData[attachFieldName]) {
                /** @type {?} */
                const attachmentId = entityData[attachFieldName]['attachmentId'];
                if (attachmentId === fileId) {
                    return true;
                }
            }
        }));
        return targetEntityData.id;
    }
    /**
     * 获取带附件的BindingList的Path
     * @private
     * @param {?} attachmentInfoFieldPath
     * @return {?}
     */
    getBindingListPathWithAttachments(attachmentInfoFieldPath) {
        /** @type {?} */
        const attachInfoBindingPath = BindingPathConverter.toBindingPathArray(attachmentInfoFieldPath);
        attachInfoBindingPath.pop();
        /** @type {?} */
        const bindingListPath = attachInfoBindingPath;
        return '/' + bindingListPath.join('/');
    }
}
if (false) {
    /**
     * 默认根目录
     * @type {?}
     * @private
     */
    AttachmentService.prototype.defaultRootDirId;
    /**
     * 视图模型
     * @type {?}
     * @private
     */
    AttachmentService.prototype.viewModelContext;
    /**
     * 附件数据服务
     * @type {?}
     * @private
     */
    AttachmentService.prototype.attachDataService;
    /**
     * 实体服务
     * @type {?}
     * @private
     */
    AttachmentService.prototype.entityService;
    /**
     * 实体服务
     * @type {?}
     * @private
     */
    AttachmentService.prototype.removeDataService;
}
export { AttachmentService };
