// @farris/mobile-command-services/esm2015/lib/attachment-services/attachment.util.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/attachment-services/attachment.util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 附件信息处理工具类
 */
class AttachmentUtil {
    /**
     * 转换为附件信息数组
     * @param {?} fileInfos
     * @return {?}
     */
    static convertToAttachmentInfos(fileInfos) {
        if (!fileInfos) {
            return [];
        }
        /** @type {?} */
        const attachmentInfos = fileInfos.map((/**
         * @param {?} fileInfo
         * @return {?}
         */
        (fileInfo) => {
            return this.convertToAttachmentInfo(fileInfo);
        }));
        return attachmentInfos;
    }
    /**
     * 转换为附件信息
     * @param {?} fileInfo
     * @return {?}
     */
    static convertToAttachmentInfo(fileInfo) {
        /** @type {?} */
        const attachmentInfo = {
            attachmentId: fileInfo.metadataId,
            fileName: fileInfo.fileName
        };
        return attachmentInfo;
    }
    /**
     * 获取附件列表中的第一个附件
     * @param {?} attachmentInfos
     * @return {?}
     */
    static getFirstAttachmentInfo(attachmentInfos) {
        if (!attachmentInfos || attachmentInfos.length === 0) {
            return;
        }
        /** @type {?} */
        const firstAttachmentInfo = attachmentInfos[0];
        return firstAttachmentInfo;
    }
    /**
     * 提取附件id数组
     * @param {?} attachmentInfos
     * @return {?}
     */
    static peekAttachmentIds(attachmentInfos) {
        if (!attachmentInfos) {
            attachmentInfos = [];
        }
        /** @type {?} */
        const attachmentIds = attachmentInfos.map((/**
         * @param {?} attachmentInfo
         * @return {?}
         */
        (attachmentInfo) => {
            return attachmentInfo.attachmentId;
        }));
        return attachmentIds;
    }
}
export { AttachmentUtil };
