// @farris/extend-file-upload/esm2015/lib/directive/f-file-select.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directive/f-file-select.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { FFileUploaderService } from '../service/f-file-uploader.service';
export class FFileSelectDirective {
    /**
     * @param {?} elementRef
     * @param {?} upload
     */
    constructor(elementRef, upload) {
        this.elementRef = elementRef;
        this.upload = upload;
        this.tDisabled = false;
        this.changeListener = null;
        this.onChange = (/**
         * @return {?}
         */
        () => {
            if (this.el.files) {
                this.upload.handleFiles(this.el.files);
                this.el.value = null;
            }
        });
        this.uploadOutput = new EventEmitter();
        this.el = this.elementRef.nativeElement;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set disabled(value) {
        if (this.tDisabled !== value) {
            this.tDisabled = value;
            if (value) {
                this.unbindEvent();
            }
            else {
                this.bindEvent();
            }
        }
    }
    /**
     * @return {?}
     */
    get disabled() {
        return this.tDisabled;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set options(value) {
        if (value) {
            this.upload.setOptions(value);
        }
    }
    ;
    // 已上传
    /**
     * @param {?} value
     * @return {?}
     */
    set uploadedCount(value) {
        this.upload.setOptions({ 'uploadedCount': value });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._sub = [];
        // 对外抛出事件
        this._sub.push(this.upload.serviceEvents.subscribe((/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            this.uploadOutput.emit(event);
        })));
        // 对内的事件
        if (this.uploadInput instanceof EventEmitter) {
            this._sub.push(this.upload.initInputEvents(this.uploadInput));
        }
        if (!this.disabled) {
            this.bindEvent();
        }
    }
    /**
     * @private
     * @return {?}
     */
    bindEvent() {
        this.changeListener = this.onChange.bind(this);
        this.el.addEventListener('change', this.changeListener);
    }
    /**
     * @private
     * @return {?}
     */
    unbindEvent() {
        this.changeListener = null;
        this.el.removeEventListener('change', this.changeListener);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.el) {
            this._sub.forEach((/**
             * @param {?} sub
             * @return {?}
             */
            sub => sub.unsubscribe()));
        }
    }
    /**
     * 重置
     * @return {?}
     */
    reset() {
        // if (this.el) {
        //   //this._sub.forEach(sub => sub.unsubscribe());
        //   //this._sub=[];
        //   this.upload.reset();
        // }    
        this.upload.reset();
    }
}
FFileSelectDirective.decorators = [
    { type: Directive, args: [{
                selector: '[FFileSelect]'
            },] }
];
/** @nocollapse */
FFileSelectDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: FFileUploaderService }
];
FFileSelectDirective.propDecorators = {
    disabled: [{ type: Input }],
    options: [{ type: Input }],
    uploadedCount: [{ type: Input }],
    uploadInput: [{ type: Input }],
    uploadOutput: [{ type: Output }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    FFileSelectDirective.prototype.tDisabled;
    /** @type {?} */
    FFileSelectDirective.prototype.uploadInput;
    /** @type {?} */
    FFileSelectDirective.prototype.uploadOutput;
    /** @type {?} */
    FFileSelectDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FFileSelectDirective.prototype._sub;
    /**
     * @type {?}
     * @private
     */
    FFileSelectDirective.prototype.changeListener;
    /** @type {?} */
    FFileSelectDirective.prototype.onChange;
    /** @type {?} */
    FFileSelectDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    FFileSelectDirective.prototype.upload;
    /* Skipping unhandled member: ;*/
}
