// @gsp-wf/wf-process-editor/esm5/lib/views/organization-potential-owner/actionOrganizationScope.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ActionOrganizationScope = /** @class */ (function () {
    function ActionOrganizationScope() {
    }
    return ActionOrganizationScope;
}());
export { ActionOrganizationScope };
if (false) {
    /** @type {?} */
    ActionOrganizationScope.prototype.enable;
    /** @type {?} */
    ActionOrganizationScope.prototype.action;
    /** @type {?} */
    ActionOrganizationScope.prototype.scope;
}
var ActionOrganizationScopeDesign = /** @class */ (function () {
    function ActionOrganizationScopeDesign() {
    }
    return ActionOrganizationScopeDesign;
}());
export { ActionOrganizationScopeDesign };
if (false) {
    /** @type {?} */
    ActionOrganizationScopeDesign.prototype.enable;
    /** @type {?} */
    ActionOrganizationScopeDesign.prototype.action;
    /** @type {?} */
    ActionOrganizationScopeDesign.prototype.scope;
}
var OrganizationPotentialOwner = /** @class */ (function () {
    function OrganizationPotentialOwner() {
    }
    return OrganizationPotentialOwner;
}());
export { OrganizationPotentialOwner };
if (false) {
    /**
     * 作用域：AssignUser, Transfer, AddSignFront, AddSignBehind
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.actionScope;
    /**
     * 部门：DEPARTMENT；单位：UNIT
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.organizationLevel;
    /**
     * 组织Id：currentDepartment, currentUnit
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.organizationId;
    /**
     * 组织编号
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.organizationCode;
    /**
     * 组织名称
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.organizationName;
    /**
     * 包含下级
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.includeChildren;
    /**
     * 遍历父级
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.parentsUntil;
    /**
     * variable 变量值
     * expression 表达式值
     * constant 常量值
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.tag;
    /**
     * 先改一下  存人员信息
     * @type {?}
     */
    OrganizationPotentialOwner.prototype.participantInfo;
}
/** @enum {string} */
var ActionScope = {
    /**
     * 指派
     */
    AssignUser: 'AssignUser',
    /**
     * 移交
     */
    Transfer: 'Transfer',
    /**
     * 征询
     */
    AddSignFront: 'AddSignFront',
    /**
     * 加签
     */
    AddSignBehind: 'AddSignBehind',
    /**
     * 知会
     */
    Notify: "Notify",
};
export { ActionScope };
/** @enum {string} */
var OrganizationLevel = {
    /**
     * 部门
     */
    DEPARTMENT: 'DEPARTMENT',
    /**
     * 单位
     */
    UNIT: 'UNIT',
    /**
     * 默认值
     */
    DEFAULT: 'DEFAULT',
    VARORGANIZATION: 'VARORGANIZATION',
};
export { OrganizationLevel };
