// @qdp/common/esm5/lib/common.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SessionService } from '@ecp-caf/caf-common';
import { RestfulService } from './service/restful-service/restful.service';
import { SchemaManagerService } from './service/schema-manager/schema-manager.service';
import { LcpService } from './service/lcp-service/lcp.service';
import { ParasMappingManagerService } from './service/paras-mapping/paras-mappingmanager.service';
import { JointSearchManagerService } from './service/joint-search/joint-searchmanager.service';
import { PrintIntegrationManagerService } from './service/print-integration/print-integrationmanager.service';
var QdpCommonModule = /** @class */ (function () {
    function QdpCommonModule() {
    }
    /**
     * @return {?}
     */
    QdpCommonModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: QdpCommonModule,
            providers: [
                SessionService,
                SchemaManagerService,
                RestfulService,
                LcpService,
                ParasMappingManagerService,
                JointSearchManagerService,
                PrintIntegrationManagerService
            ]
        };
    };
    QdpCommonModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                    ],
                    declarations: [],
                    exports: []
                },] }
    ];
    return QdpCommonModule;
}());
export { QdpCommonModule };
