// @gsp-lcm/metadata-selector/esm2015/metadata/metadata-select/components/treeclassifier/treeclassifier.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/treeclassifier/treeclassifier.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { MetadataService } from '../../metadata.service';
import { GSPMetadataService } from '../../../metadata.service';
export class TreeclassifierComponent {
    /**
     * @param {?} metadataService
     * @param {?} gspMetadataService
     */
    constructor(metadataService, gspMetadataService) {
        this.metadataService = metadataService;
        this.gspMetadataService = gspMetadataService;
        this.selectedKeys = ['0'];
        this.classifier = [
            { id: 'project', text: '当前工程' }
        ];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.metadataService.GetMetadataProjPath(this.metadataPath).subscribe((/**
         * @param {?} proj
         * @return {?}
         */
        proj => {
            this.metadataService.currentProjectPath = proj.projectPath;
            if (this.onlyDisplayCurrentProject === true) {
                this.files = this.classifier;
                this.metadataService.selectedChange.emit(this.files[0]); // 默认本工程下的
            }
            else {
                // 树节点mock数据
                this.gspMetadataService.getMavenSetting().subscribe((/**
                 * @param {?} maven
                 * @return {?}
                 */
                maven => {
                    if (maven === true) {
                        this.gspMetadataService.getCodeLanguage(proj.projectPath).subscribe((/**
                         * @param {?} lang
                         * @return {?}
                         */
                        lang => {
                            if (lang === 'java') {
                                this.files = this.metadataService.getMetadataClassifier('bo');
                                this.metadataService.selectedChange.emit(this.files[0]); // 默认本工程下的
                            }
                            else {
                                this.files = this.metadataService.getMetadataClassifier('project');
                                this.metadataService.selectedChange.emit(this.files[0]); // 默认本工程下的
                            }
                        }));
                    }
                    else {
                        this.files = this.metadataService.getMetadataClassifier('project');
                        this.metadataService.selectedChange.emit(this.files[0]); // 默认本工程下的
                    }
                }));
            }
        }));
    }
    // 树节点点击后事件
    /**
     * @param {?} event
     * @return {?}
     */
    nodeSelect(event) {
        this.metadataService.selectedChange.emit(event.dataItem);
    }
}
TreeclassifierComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'app-treeclassifier',
                template: "<div class=\"templateTree befTree\">\r\n    <div class=\"treeClassifier\">\r\n        <kendo-treeview\r\n            [nodes]=\"files\"\r\n            textField=\"text\"\r\n            kendoTreeViewExpandable\r\n            kendoTreeViewSelectable\r\n            kendoTreeViewHierarchyBinding\r\n            childrenField=\"items\"\r\n            (selectionChange)=\"this.nodeSelect($event)\"\r\n            [(selectedKeys)]=\"this.selectedKeys\"\r\n        >\r\n        </kendo-treeview>\r\n    </div>\r\n</div>",
                styles: [".treeClassifier .k-treeview{overflow:visible}.templateTree{padding:5px;margin-left:5px}"]
            }] }
];
/** @nocollapse */
TreeclassifierComponent.ctorParameters = () => [
    { type: MetadataService },
    { type: GSPMetadataService }
];
TreeclassifierComponent.propDecorators = {
    metadataPath: [{ type: Input }],
    onlyDisplayCurrentProject: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TreeclassifierComponent.prototype.metadataPath;
    /** @type {?} */
    TreeclassifierComponent.prototype.onlyDisplayCurrentProject;
    /** @type {?} */
    TreeclassifierComponent.prototype.files;
    /** @type {?} */
    TreeclassifierComponent.prototype.selectedKeys;
    /** @type {?} */
    TreeclassifierComponent.prototype.classifier;
    /**
     * @type {?}
     * @private
     */
    TreeclassifierComponent.prototype.metadataService;
    /**
     * @type {?}
     * @private
     */
    TreeclassifierComponent.prototype.gspMetadataService;
}
