// @farris/mobile-bef/esm5/lib/bef-lookup-data.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef-lookup-data.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { map } from 'rxjs/operators';
import { HttpMethods } from '@farris/mobile-devkit';
import { RequestInfoUtil } from './utils/index';
/**
 * Bef帮助取数服务
 */
var /**
 * Bef帮助取数服务
 */
BefLookupDataService = /** @class */ (function () {
    /**
     * 构造函数
     */
    function BefLookupDataService(viewModelContext) {
        this.viewModelContext = viewModelContext;
        this.befRepository = (/** @type {?} */ (this.viewModelContext.repository));
    }
    /**
     * 帮助取数
     */
    /**
     * 帮助取数
     * @param {?} helpMetadataId
     * @param {?=} queryParam
     * @return {?}
     */
    BefLookupDataService.prototype.getData = /**
     * 帮助取数
     * @param {?} helpMetadataId
     * @param {?=} queryParam
     * @return {?}
     */
    function (helpMetadataId, queryParam) {
        /** @type {?} */
        var tableName = helpMetadataId.split('.')[0];
        /** @type {?} */
        var labelId = helpMetadataId.split('.')[1];
        queryParam = queryParam || {};
        return this.extendGetHelpData(labelId, tableName, queryParam);
    };
    /**
     * 扩展的帮助取数
     */
    /**
     * 扩展的帮助取数
     * @private
     * @param {?} labelId
     * @param {?} tableName
     * @param {?} queryParam
     * @return {?}
     */
    BefLookupDataService.prototype.extendGetHelpData = /**
     * 扩展的帮助取数
     * @private
     * @param {?} labelId
     * @param {?} tableName
     * @param {?} queryParam
     * @return {?}
     */
    function (labelId, tableName, queryParam) {
        /** @type {?} */
        var url = this.befRepository.apiProxy.baseUrl + "/extension/elementhelps";
        /** @type {?} */
        var requestInfo = RequestInfoUtil.buildRequestInfo(this.befRepository);
        /** @type {?} */
        var body = {
            labelId: labelId,
            nodeCode: tableName,
            queryParam: queryParam,
            requestInfo: requestInfo
        };
        /** @type {?} */
        var requestConfig = {
            body: body
        };
        /** @type {?} */
        var result$ = this.befRepository.apiProxy.request(HttpMethods.PUT, url, requestConfig);
        return result$.pipe(map((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            return result;
        })));
    };
    return BefLookupDataService;
}());
/**
 * Bef帮助取数服务
 */
export { BefLookupDataService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    BefLookupDataService.prototype.viewModelContext;
    /**
     * @type {?}
     * @private
     */
    BefLookupDataService.prototype.befRepository;
    /** @type {?} */
    BefLookupDataService.prototype.context;
}
