// @farris/mobile-bef/esm5/lib/bef_proxy_extend.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_proxy_extend.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { of, EMPTY } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { RequestInfoUtil } from './utils';
import { BE_ERROR_HANDLER__TOKEN } from './tokens';
/**
 * 代理钩子
 * @record
 */
export function IProxyExtend() { }
if (false) {
    /**
     * @param {?} body
     * @return {?}
     */
    IProxyExtend.prototype.extendBody = function (body) { };
    /**
     * @param {?} headers
     * @return {?}
     */
    IProxyExtend.prototype.extendHeaders = function (headers) { };
    /**
     * @param {?} response
     * @param {?=} ignoreChanges
     * @return {?}
     */
    IProxyExtend.prototype.onResponse = function (response, ignoreChanges) { };
    /**
     * @param {?} error
     * @param {?} selfHandError
     * @param {?} ignoreError
     * @return {?}
     */
    IProxyExtend.prototype.onError = function (error, selfHandError, ignoreError) { };
}
/**
 * Bef代理扩展类
 */
var /**
 * Bef代理扩展类
 */
BefProxyExtend = /** @class */ (function () {
    /**
     * 构造函数
     */
    function BefProxyExtend(repository) {
        this.repository = repository;
    }
    Object.defineProperty(BefProxyExtend.prototype, "injector", {
        /**
       * 注入器
       */
        get: /**
         * 注入器
         * @private
         * @return {?}
         */
        function () {
            return this.repository.getInjector();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 返回结果处理
     */
    /**
     * 返回结果处理
     * @param {?} response
     * @return {?}
     */
    BefProxyExtend.prototype.onResponse = /**
     * 返回结果处理
     * @param {?} response
     * @return {?}
     */
    function (response) {
        this.handleResponseHeaders(response.headers);
        return this.handleResponseBody(response.body);
    };
    /**
     * 处理服务器端返回的headers数据
     */
    /**
     * 处理服务器端返回的headers数据
     * @private
     * @param {?} headers
     * @return {?}
     */
    BefProxyExtend.prototype.handleResponseHeaders = /**
     * 处理服务器端返回的headers数据
     * @private
     * @param {?} headers
     * @return {?}
     */
    function (headers) {
        this.repository.sessionService.handleResponseHeaders(headers);
    };
    /**
     * 处理服务器端返回的的body数据
     */
    /**
     * 处理服务器端返回的的body数据
     * @private
     * @param {?} responseInfo
     * @return {?}
     */
    BefProxyExtend.prototype.handleResponseBody = /**
     * 处理服务器端返回的的body数据
     * @private
     * @param {?} responseInfo
     * @return {?}
     */
    function (responseInfo) {
        if (responseInfo && responseInfo.innerDataChange) {
            this.repository.entityManager.handleDataChangeDetails(responseInfo.innerDataChange);
        }
        this.repository.entityManager.clearAllEntityChanges();
        if (responseInfo && responseInfo.hasOwnProperty('returnValue')) {
            return responseInfo.returnValue;
        }
        else {
            return responseInfo;
        }
    };
    /**
     * 错误处理
     */
    /**
     * 错误处理
     * @param {?} error
     * @return {?}
     */
    BefProxyExtend.prototype.onError = /**
     * 错误处理
     * @param {?} error
     * @return {?}
     */
    function (error) {
        var e_1, _a;
        // 获取所有load组件实例，消除
        /** @type {?} */
        var loadingServices = window['DEVKIT_LOADING_SERVICE'];
        if (loadingServices && loadingServices instanceof Array && loadingServices.length > 0) {
            try {
                for (var loadingServices_1 = tslib_1.__values(loadingServices), loadingServices_1_1 = loadingServices_1.next(); !loadingServices_1_1.done; loadingServices_1_1 = loadingServices_1.next()) {
                    var loadingService = loadingServices_1_1.value;
                    if (typeof (loadingService.hide) === 'function') {
                        loadingService.hide();
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (loadingServices_1_1 && !loadingServices_1_1.done && (_a = loadingServices_1.return)) _a.call(loadingServices_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        // 捕获到异常，处理异常信息
        /** @type {?} */
        var exceptionService = this.injector.get(BE_ERROR_HANDLER__TOKEN, null);
        if (!!exceptionService) {
            exceptionService.show(error);
            return EMPTY;
        }
        else {
            return EMPTY;
        }
    };
    /**
     * 扩展Headers
     */
    /**
     * 扩展Headers
     * @param {?} headers
     * @return {?}
     */
    BefProxyExtend.prototype.extendHeaders = /**
     * 扩展Headers
     * @param {?} headers
     * @return {?}
     */
    function (headers) {
        var _this = this;
        /** @type {?} */
        var $getSessionId = this.repository.sessionService.getBeSessionId();
        return $getSessionId.pipe(switchMap((/**
         * @return {?}
         */
        function () {
            headers = _this.repository.sessionService.extendRequestHeaders(headers);
            return of(headers);
        })));
    };
    /**
     * 扩展Body
     */
    /**
     * 扩展Body
     * @param {?} body
     * @return {?}
     */
    BefProxyExtend.prototype.extendBody = /**
     * 扩展Body
     * @param {?} body
     * @return {?}
     */
    function (body) {
        return RequestInfoUtil.appendRequestInfoToBody(body, this.repository);
    };
    return BefProxyExtend;
}());
/**
 * Bef代理扩展类
 */
export { BefProxyExtend };
if (false) {
    /**
     * @type {?}
     * @private
     */
    BefProxyExtend.prototype.repository;
}
