// @farris/mobile-bef/esm5/lib/bef_data_service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_data_service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { of, zip, EMPTY } from 'rxjs';
import { tap, map, switchMap } from 'rxjs/operators';
import { DataChangeType, } from '@farris/mobile-devkit';
/**
 * Bef数据访问服务
 * @template T
 */
var /**
 * Bef数据访问服务
 * @template T
 */
BefDataService = /** @class */ (function () {
    /**
     * 构造函数
     */
    function BefDataService(repository) {
        this.repository = repository;
    }
    Object.defineProperty(BefDataService.prototype, "apiProxy", {
        /**
         * Api代理类
         */
        get: /**
         * Api代理类
         * @private
         * @return {?}
         */
        function () {
            return this.repository.apiProxy;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 获取实体集合
     */
    /**
     * 获取实体集合
     * @param {?} filter
     * @param {?} sort
     * @param {?} pageSize
     * @param {?} pageIndex
     * @return {?}
     */
    BefDataService.prototype.getList = /**
     * 获取实体集合
     * @param {?} filter
     * @param {?} sort
     * @param {?} pageSize
     * @param {?} pageIndex
     * @return {?}
     */
    function (filter, sort, pageSize, pageIndex) {
        var _this = this;
        /** @type {?} */
        var entityFilter = this.buildEntityFilter(filter, sort, pageSize, pageIndex);
        /** @type {?} */
        var query$ = this.apiProxy.extendQuery(entityFilter);
        /** @type {?} */
        var result$ = query$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        function (returnValue) {
            _this.repository.setPaginationInfo(((/** @type {?} */ (returnValue))).pagination);
            /** @type {?} */
            var listData = ((/** @type {?} */ (returnValue))).result;
            /** @type {?} */
            var entities = _this.repository.buildEntities(listData);
            if (_this.determineIsAppend(pageIndex)) {
                _this.repository.entityCollection.addEntities(entities);
            }
            else {
                _this.repository.entityCollection.loadEntities(entities);
            }
            return entities;
        })));
        return result$;
    };
    /**
     * 根据当前页判断是否追加
     */
    /**
     * 根据当前页判断是否追加
     * @param {?} pageIndex
     * @return {?}
     */
    BefDataService.prototype.determineIsAppend = /**
     * 根据当前页判断是否追加
     * @param {?} pageIndex
     * @return {?}
     */
    function (pageIndex) {
        return pageIndex > 1;
    };
    /**
     * 获取单个实体
     */
    /**
     * 获取单个实体
     * @param {?} id
     * @return {?}
     */
    BefDataService.prototype.getById = /**
     * 获取单个实体
     * @param {?} id
     * @return {?}
     */
    function (id) {
        var _this = this;
        /** @type {?} */
        var retrieve$ = this.apiProxy.extendRetrieve(id);
        /** @type {?} */
        var result$ = retrieve$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        function (returnValue) {
            /** @type {?} */
            var entityData = returnValue;
            /** @type {?} */
            var entity = _this.repository.buildEntity(entityData);
            _this.repository.entityCollection.loadEntities([entity]);
            return entity;
        })));
        return result$;
    };
    /**
     * 更新并对实体加锁
     */
    /**
     * 更新并对实体加锁
     * @param {?} id
     * @return {?}
     */
    BefDataService.prototype.editById = /**
     * 更新并对实体加锁
     * @param {?} id
     * @return {?}
     */
    function (id) {
        var _this = this;
        /** @type {?} */
        var entity = this.repository.entityCollection.getEntityById(id);
        if (!entity) {
            return of(null);
        }
        /** @type {?} */
        var retrieve$ = this.apiProxy.edit(id);
        /** @type {?} */
        var result$ = retrieve$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        function (returnValue) {
            /** @type {?} */
            var entityData = returnValue.data;
            /** @type {?} */
            var entityToUpdate = _this.repository.entityCollection.getEntityById(id);
            if (entityToUpdate && entityData) {
                _this.reloadEntityData(entityToUpdate, entityData);
            }
            return entityToUpdate;
        })));
        return result$;
    };
    /**
     * 根据id更新实体
     * @param id 实体id
     */
    /**
     * 根据id更新实体
     * @param {?} id 实体id
     * @return {?}
     */
    BefDataService.prototype.updateById = /**
     * 根据id更新实体
     * @param {?} id 实体id
     * @return {?}
     */
    function (id) {
        var _this = this;
        /** @type {?} */
        var entity = this.repository.entityCollection.getEntityById(id);
        if (!entity) {
            return EMPTY;
        }
        /** @type {?} */
        var retrieve$ = this.apiProxy.extendRetrieve(id);
        /** @type {?} */
        var result$ = retrieve$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        function (returnValue) {
            /** @type {?} */
            var entityData = returnValue;
            /** @type {?} */
            var entityToUpdate = _this.repository.entityCollection.getEntityById(id);
            _this.reloadEntityData(entityToUpdate, entityData);
            return entityToUpdate;
        })));
        return result$;
    };
    /**
     * 重新加载实体数据
     */
    /**
     * 重新加载实体数据
     * @private
     * @param {?} entity
     * @param {?} entityData
     * @return {?}
     */
    BefDataService.prototype.reloadEntityData = /**
     * 重新加载实体数据
     * @private
     * @param {?} entity
     * @param {?} entityData
     * @return {?}
     */
    function (entity, entityData) {
        if (!entity) {
            return;
        }
        entity.load(entityData);
        entity.changes.splice(0, entity.changes.length);
    };
    /**
     * 创建新实体，并加载
     */
    /**
     * 创建新实体，并加载
     * @param {?=} defaultValue
     * @return {?}
     */
    BefDataService.prototype.create = /**
     * 创建新实体，并加载
     * @param {?=} defaultValue
     * @return {?}
     */
    function (defaultValue) {
        var _this = this;
        /** @type {?} */
        var result$ = this.apiProxy.create(defaultValue);
        return result$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        function (returnValue) {
            /** @type {?} */
            var newEntityData = returnValue;
            /** @type {?} */
            var newEntity = _this.repository.buildEntity(newEntityData);
            _this.repository.entityCollection.loadEntities([newEntity]);
            _this.repository.dataChangeHistory.addChange({ dataId: newEntity.primaryValue, changeType: DataChangeType.Add });
            return newEntity;
        })));
    };
    /**
     * 追加实体
     */
    /**
     * 追加实体
     * @param {?=} defaultValue
     * @return {?}
     */
    BefDataService.prototype.append = /**
     * 追加实体
     * @param {?=} defaultValue
     * @return {?}
     */
    function (defaultValue) {
        var _this = this;
        /** @type {?} */
        var append$ = this.apiProxy.create(defaultValue);
        return append$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        function (returnValue) {
            /** @type {?} */
            var newEntityData = returnValue;
            /** @type {?} */
            var newEntity = _this.repository.buildEntity(newEntityData);
            _this.repository.entityCollection.addEntity(newEntity);
            return newEntity;
        })));
    };
    /**
     * 创建子实体，并追加
     * @param fpath 父路径，格式形如：/1/edus（从表）或/1/edus/2/grades（从从表）
     */
    /**
     * 创建子实体，并追加
     * @param {?} fpath 父路径，格式形如：/1/edus（从表）或/1/edus/2/grades（从从表）
     * @return {?}
     */
    BefDataService.prototype.appendByPath = /**
     * 创建子实体，并追加
     * @param {?} fpath 父路径，格式形如：/1/edus（从表）或/1/edus/2/grades（从从表）
     * @return {?}
     */
    function (fpath) {
        var _this = this;
        /** @type {?} */
        var append$ = this.apiProxy.createByPath(fpath);
        return append$.pipe(map((/**
         * @param {?} returnValue
         * @return {?}
         */
        function (returnValue) {
            /** @type {?} */
            var newEntityData = returnValue;
            /** @type {?} */
            var newEntity = _this.repository.entityManager.appendEntityByPath(fpath, newEntityData, newEntityData);
            return newEntity;
        })));
    };
    /**
     * 根据id删除实体
     * @param id 内码
     */
    /**
     * 根据id删除实体
     * @param {?} id 内码
     * @param {?=} ifSave
     * @return {?}
     */
    BefDataService.prototype.removeById = /**
     * 根据id删除实体
     * @param {?} id 内码
     * @param {?=} ifSave
     * @return {?}
     */
    function (id, ifSave) {
        var _this = this;
        ifSave = (ifSave === undefined) ? true : ifSave;
        /** @type {?} */
        var delete$ = this.apiProxy.extendDelete(id);
        // 从本地实体集合中移除
        return delete$.pipe(switchMap((/**
         * @return {?}
         */
        function () {
            // 执行保存
            if (ifSave === true) {
                return _this.applyChangesById(id).pipe(tap((/**
                 * @param {?} result
                 * @return {?}
                 */
                function (result) {
                    if (result) {
                        _this.repository.entityCollection.removeEntityById(id);
                    }
                })));
            }
            else {
                // 从实体集合中删除
                _this.repository.entityCollection.removeEntityById(id);
                _this.repository.dataChangeHistory.addChange({ dataId: id, changeType: DataChangeType.Delete });
                return of(true);
            }
        })));
    };
    /**
     * 删除并保存
     * @param id 要删除的实体id
     */
    /**
     * 删除并保存
     * @param {?} id 要删除的实体id
     * @return {?}
     */
    BefDataService.prototype.removeAndSaveById = /**
     * 删除并保存
     * @param {?} id 要删除的实体id
     * @return {?}
     */
    function (id) {
        var _this = this;
        /** @type {?} */
        var delete$ = this.apiProxy.deleteAndSave(id);
        /** @type {?} */
        var result$ = delete$.pipe(switchMap((/**
         * @return {?}
         */
        function () {
            _this.repository.entityCollection.removeEntityById(id);
            return of(true);
        })));
        return result$;
    };
    /**
     * 批量删除
     * @param id 内码
     */
    /**
     * 批量删除
     * @param {?} ids
     * @return {?}
     */
    BefDataService.prototype.removeByIds = /**
     * 批量删除
     * @param {?} ids
     * @return {?}
     */
    function (ids) {
        var _this = this;
        // 服务器端删除
        /** @type {?} */
        var delete$ = this.apiProxy.extendBatchDelete(ids);
        // 从本地实体集合中移除
        return delete$.pipe(switchMap((/**
         * @return {?}
         */
        function () {
            // 执行保存
            return _this.applyChangesByIdArray(ids).pipe(tap((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                if (result) {
                    // 从实体集合中删除
                    _this.repository.entityCollection.removeEntities((/**
                     * @param {?} value
                     * @return {?}
                     */
                    function (value) {
                        return _this.checkEntityValueExists(value, ids);
                    }));
                }
            })));
        })));
    };
    /**
     * 检测实体的主键值是否存在于数组中，如果存在返回true
     * @param array 待遍历数组
     */
    /**
     * 检测实体的主键值是否存在于数组中，如果存在返回true
     * @private
     * @param {?} entity
     * @param {?} primaryValueArray
     * @return {?}
     */
    BefDataService.prototype.checkEntityValueExists = /**
     * 检测实体的主键值是否存在于数组中，如果存在返回true
     * @private
     * @param {?} entity
     * @param {?} primaryValueArray
     * @return {?}
     */
    function (entity, primaryValueArray) {
        /** @type {?} */
        var isExistInTargetArray = false;
        for (var i = 0; i < primaryValueArray.length; i++) {
            if (entity.primaryValue === primaryValueArray[i]) {
                isExistInTargetArray = true;
                break;
            }
        }
        return isExistInTargetArray;
    };
    /**
     * 删除子级
     * @param path 路径
     * @param id   内码
     */
    /**
     * 删除子级
     * @param {?} fpath
     * @param {?} id   内码
     * @return {?}
     */
    BefDataService.prototype.removeByPath = /**
     * 删除子级
     * @param {?} fpath
     * @param {?} id   内码
     * @return {?}
     */
    function (fpath, id) {
        var _this = this;
        /** @type {?} */
        var delete$ = this.apiProxy.extendDeletByPath(fpath, id);
        // 从本地实体集合中移除
        return delete$.pipe(map((/**
         * @return {?}
         */
        function () {
            _this.repository.entityManager.removeEntityByPath(fpath, id);
            _this.repository.dataChangeHistory.addChange({ fpath: fpath, dataId: id, changeType: DataChangeType.Delete });
            return true;
        })));
    };
    /**
     * 将id对应的实体的变更提交的服务器端
     */
    /**
     * 将id对应的实体的变更提交的服务器端
     * @param {?} id
     * @return {?}
     */
    BefDataService.prototype.updateChangesById = /**
     * 将id对应的实体的变更提交的服务器端
     * @param {?} id
     * @return {?}
     */
    function (id) {
        var _this = this;
        /** @type {?} */
        var entity = this.repository.entityCollection.getEntityById(id);
        // 变更集为空
        if (!entity.changes || entity.changes.length === 0) {
            return of(true);
        }
        /** @type {?} */
        var changeDetail = this.repository.entityManager.buildEntityChangeDetailById(id);
        /** @type {?} */
        var update$ = this.apiProxy.update(changeDetail);
        /** @type {?} */
        var result$ = update$.pipe(tap((/**
         * @return {?}
         */
        function () {
            _this.repository.entityManager.clearEntityChangesById(id);
        })), map((/**
         * @return {?}
         */
        function () {
            return true;
        })));
        return result$;
    };
    /**
     * 提交所有变更
     */
    /**
     * 提交所有变更
     * @return {?}
     */
    BefDataService.prototype.updateAllChanges = /**
     * 提交所有变更
     * @return {?}
     */
    function () {
        var _this = this;
        // 遍历实体，提交变更
        /** @type {?} */
        var updateResults = [];
        /** @type {?} */
        var entities = this.repository.entityCollection.toArray();
        if (entities.length === 0) {
            return of(true);
        }
        entities.forEach((/**
         * @param {?} entity
         * @return {?}
         */
        function (entity) {
            /** @type {?} */
            var updateResult$ = _this.updateChangesById(entity.primaryValue);
            updateResults.push(updateResult$);
        }));
        // 串联流
        /** @type {?} */
        var result$ = zip.apply(void 0, tslib_1.__spread(updateResults)).pipe(map((/**
         * @return {?}
         */
        function () {
            return true;
        })));
        return result$;
    };
    /**
     * 应用变更
     */
    /**
     * 应用变更
     * @return {?}
     */
    BefDataService.prototype.applyChanges = /**
     * 应用变更
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var save$ = this.apiProxy.save();
        /** @type {?} */
        var result$ = save$.pipe(tap((/**
         * @return {?}
         */
        function () {
            _this.repository.entityManager.clearAllEntityChanges();
            _this.repository.dataChangeHistory.clear();
        })), map((/**
         * @return {?}
         */
        function () {
            return true;
        })));
        return result$;
    };
    /**
     * 应用idArray对应数据的变更
     */
    /**
     * 应用idArray对应数据的变更
     * @param {?} ids
     * @return {?}
     */
    BefDataService.prototype.applyChangesByIdArray = /**
     * 应用idArray对应数据的变更
     * @param {?} ids
     * @return {?}
     */
    function (ids) {
        var _this = this;
        /** @type {?} */
        var save$ = this.apiProxy.save();
        /** @type {?} */
        var result$ = save$.pipe(tap((/**
         * @return {?}
         */
        function () {
            _this.repository.entityManager.clearEntityChangesByIds(ids);
            _this.repository.dataChangeHistory.clearByIds(ids);
        })), map((/**
         * @return {?}
         */
        function () {
            return true;
        })));
        return result$;
    };
    /**
     * 应用id对应数据的变更
     */
    /**
     * 应用id对应数据的变更
     * @param {?} id
     * @return {?}
     */
    BefDataService.prototype.applyChangesById = /**
     * 应用id对应数据的变更
     * @param {?} id
     * @return {?}
     */
    function (id) {
        var _this = this;
        /** @type {?} */
        var save$ = this.apiProxy.save();
        /** @type {?} */
        var result$ = save$.pipe(tap((/**
         * @return {?}
         */
        function () {
            _this.repository.entityManager.clearEntityChangesById(id);
            _this.repository.dataChangeHistory.clearByIds([id]);
        })), map((/**
         * @return {?}
         */
        function () {
            return true;
        })));
        return result$;
    };
    /**
     * 取消变更
     */
    /**
     * 取消变更
     * @return {?}
     */
    BefDataService.prototype.cancelChanges = /**
     * 取消变更
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var cancel$ = this.apiProxy.cancel();
        /** @type {?} */
        var result$ = cancel$.pipe(tap((/**
         * @return {?}
         */
        function () {
            _this.repository.entityManager.clearAllEntityChanges();
            _this.repository.dataChangeHistory.clear();
        })), map((/**
         * @return {?}
         */
        function () {
            return true;
        })));
        return result$;
    };
    /**
     * 构造EntityFilter
     */
    /**
     * 构造EntityFilter
     * @private
     * @param {?} filter
     * @param {?} sort
     * @param {?} pageSize
     * @param {?} pageIndex
     * @return {?}
     */
    BefDataService.prototype.buildEntityFilter = /**
     * 构造EntityFilter
     * @private
     * @param {?} filter
     * @param {?} sort
     * @param {?} pageSize
     * @param {?} pageIndex
     * @return {?}
     */
    function (filter, sort, pageSize, pageIndex) {
        /** @type {?} */
        var entityFilter = {
            FilterConditions: filter,
            SortConditions: sort,
            IsUsePagination: pageSize === 0 ? false : true,
            Pagination: {
                PageIndex: pageIndex,
                PageSize: pageSize,
                PageCount: 0,
                TotalCount: 0
            }
        };
        return entityFilter;
    };
    return BefDataService;
}());
if (false) {
    /**
     * 实体仓库
     * @type {?}
     * @private
     */
    BefDataService.prototype.repository;
}
export { BefDataService };
