// @gsp-cmp/common-component/esm2015/convertor/util/ref-structure-util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { RefStructureInfo } from '../../define/ref-structure-info';
export class RefStructureUtil {
    /**
     * @param {?} refInfo
     * @return {?}
     */
    static convertToObject(refInfo) {
        /** @type {?} */
        let result = new Object();
        result["id"] = refInfo.id;
        result["referId"] = refInfo.referId;
        result["referType"] = refInfo.referType;
        result["referCode"] = refInfo.referCode;
        result["structuredTypeId"] = refInfo.structuredTypeId;
        result["code"] = refInfo.code;
        result["name"] = refInfo.name;
        return result;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    static initFromObject(obj) {
        /** @type {?} */
        let result = new RefStructureInfo();
        result.id = obj['id'];
        result.referId = obj['referId'];
        result.referType = obj['referType'];
        result.referCode = obj['referCode'];
        result.structuredTypeId = obj['structuredTypeId'];
        result.code = obj['code'];
        result.name = obj['name'];
        return result;
    }
}
