// @farris/ui-search-box/esm5/lib/search-fields/editors/number/flexible-number.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Component, Input, Injector, ChangeDetectorRef, Output } from '@angular/core';
var SearchFieldFlexibleNumberComponent = /** @class */ (function () {
    function SearchFieldFlexibleNumberComponent(injecotr, cd) {
        this.injecotr = injecotr;
        this.cd = cd;
        this.valueChange = new EventEmitter();
        // 是否为区间日期
        this.single = true;
    }
    /**
     * @return {?}
     */
    SearchFieldFlexibleNumberComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.single = this.field.control.single;
        if (this.field) {
            if (this.field.value) {
                var _a = this.field.value, value = _a.value, begin = _a.begin, end = _a.end;
                this.value = value;
                this.beginValue = begin;
                this.endValue = end;
            }
            this.bigNumber = this.field.control.bigNumber;
            this.precision = this.field.control.precision || 0;
            if (!this.bigNumber) {
                this.min = -2147483648;
                this.max = 2147483647;
            }
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SearchFieldFlexibleNumberComponent.prototype.onSingleValueChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.single = !$event;
        this.valueChange.emit({ field: this.field.code, value: { value: this.value, end: this.endValue, begin: this.beginValue, single: this.single } });
        this.cd.detectChanges();
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SearchFieldFlexibleNumberComponent.prototype.onValueChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.valueChange.emit({ field: this.field.code, value: { value: $event, single: true } });
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SearchFieldFlexibleNumberComponent.prototype.onBeginValueChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.valueChange.emit({ field: this.field.code, value: { begin: $event, end: this.endValue, single: false } });
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SearchFieldFlexibleNumberComponent.prototype.onEndValueChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.valueChange.emit({ field: this.field.code, value: { begin: this.beginValue, end: $event, single: false } });
    };
    SearchFieldFlexibleNumberComponent.decorators = [
        { type: Component, args: [{
                    selector: 'search-filed-editor-number',
                    template: "<div class=\"f-panel-filter-tip\">\r\n    <span class=\"panel-filter-tip-text\">{{ 'filterPanel.intervalFilter' | locale }}</span>\r\n    <farris-switch size=\"small\" [checked]=\"!single\" (valueChange)=\"onSingleValueChange($event)\"></farris-switch>\r\n</div>\r\n<div class=\"f-panel-filter-content\">\r\n    <ng-container *ngIf=\"single; else elseTemplate\">\r\n        <farris-number-spinner\r\n            class=\"f-utils-fill-flex-row\"\r\n            [(ngModel)]=\"value\"\r\n            [precision]=\"precision\"\r\n            [bigNumber]=\"bigNumber\"\r\n            [canNull]=\"true\"\r\n            [min]=\"min\" [max]=\"max\"\r\n            [placeholder]=\"placeholder\"\r\n            (valueChange)=\"onValueChange($event)\"\r\n            #numberText\r\n        ></farris-number-spinner>\r\n    </ng-container>\r\n    <ng-template #elseTemplate>\r\n        <div class=\"d-flex panel-flexible-range-wrapper\">\r\n            <farris-number-spinner\r\n                [(ngModel)]=\"beginValue\"\r\n                [precision]=\"precision\"\r\n                [bigNumber]=\"bigNumber\"\r\n                [canNull]=\"true\"\r\n                [min]=\"min\" [max]=\"endValue?endValue:max\"\r\n                [placeholder]=\"beginPlaceHolder\"\r\n                (valueChange)=\"onBeginValueChange($event)\"\r\n                #beginNumberText\r\n            ></farris-number-spinner>\r\n            <span class=\"f-icon f-icon-orientation-arrow\"></span>\r\n            <farris-number-spinner\r\n                [(ngModel)]=\"endValue\"\r\n                [precision]=\"precision\"\r\n                [bigNumber]=\"bigNumber\"\r\n                [canNull]=\"true\"\r\n                [min]=\"beginValue?beginValue:min\" [max]=\"max\"\r\n                [placeholder]=\"endPlaceHolder\"\r\n                (valueChange)=\"onEndValueChange($event)\"\r\n            ></farris-number-spinner>\r\n        </div>\r\n    </ng-template>\r\n    \r\n</div>"
                }] }
    ];
    /** @nocollapse */
    SearchFieldFlexibleNumberComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ChangeDetectorRef }
    ]; };
    SearchFieldFlexibleNumberComponent.propDecorators = {
        field: [{ type: Input }],
        valueChange: [{ type: Output }]
    };
    return SearchFieldFlexibleNumberComponent;
}());
export { SearchFieldFlexibleNumberComponent };
if (false) {
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.field;
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.valueChange;
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.single;
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.value;
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.precision;
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.bigNumber;
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.placeholder;
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.beginValue;
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.endValue;
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.beginPlaceHolder;
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.endPlaceHolder;
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.max;
    /** @type {?} */
    SearchFieldFlexibleNumberComponent.prototype.min;
    /**
     * @type {?}
     * @private
     */
    SearchFieldFlexibleNumberComponent.prototype.injecotr;
    /**
     * @type {?}
     * @private
     */
    SearchFieldFlexibleNumberComponent.prototype.cd;
}
