// @gsp-svc/cloudprint/esm2015/lib/utils/cloudprint.utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// @dynamic
export class Utils {
    /**
     * 判断是否PC端 or 移动端
     * true:PC
     * false:Mobile
     * @return {?}
     */
    static isPC() {
        if ((navigator.userAgent.match(/(phone|pad|pod|iPhone|iPod|ios|iPad|Android|Mobile|BlackBerry|IEMobile|MQQBrowser|JUC|Fennec|wOSBrowser|BrowserNG|WebOS|Symbian|Windows Phone)/i))) {
            return false; // 移动端
        }
        else {
            return true;
        }
    }
    /**
     * @return {?}
     */
    static uuid() {
        return [this.gen(2), this.gen(1), this.gen(1), this.gen(1), this.gen(3)].join('-');
    }
    /**
     * @param {?} rendedFormat
     * @param {?} ele
     * @return {?}
     */
    static renderPdfByByte(rendedFormat, ele) {
        if (rendedFormat) {
            /** @type {?} */
            const lock = setInterval((/**
             * @return {?}
             */
            function () {
                if (ele) {
                    /** @type {?} */
                    const v = (/** @type {?} */ (document.getElementById('viewframe')));
                    if (v) {
                        /** @type {?} */
                        const filebtn = v.contentWindow.document.getElementById('openFile');
                        /** @type {?} */
                        const printbtn = v.contentWindow.document.getElementById('print');
                        /** @type {?} */
                        const downbtn = v.contentWindow.document.getElementById('download');
                        /** @type {?} */
                        let file = false;
                        /** @type {?} */
                        let down = false;
                        /** @type {?} */
                        let print = false;
                        if (filebtn) {
                            filebtn.style.display = 'none';
                            file = true;
                        }
                        if (downbtn) {
                            downbtn.style.display = 'none';
                            down = true;
                        }
                        if (printbtn) {
                            printbtn.style.display = 'none';
                            print = true;
                        }
                        if (file && down && print) {
                            clearInterval(lock);
                        }
                    }
                    else {
                        /** @type {?} */
                        const binaryData = [];
                        // binaryData.push(self.convertBase64ToUint8Array(rendedFormat));
                        binaryData.push(new Uint8Array(rendedFormat));
                        this.url = window.URL.createObjectURL(new Blob(binaryData, { type: 'application/pdf' }));
                        // 通过pdfjs来预览pdf
                        ele.innerHTML = '<iframe id="viewframe" src="/platform/common/web/pdfjs/viewer.html?file=' + encodeURIComponent(this.url) + '" height="100%" width="100%"></iframe>';
                        ele.hidden = false;
                    }
                }
            }), 200);
        }
    }
    /**
     * @private
     * @param {?} count
     * @return {?}
     */
    static gen(count) {
        /** @type {?} */
        let out = '';
        for (let i = 0; i < count; i++) {
            // tslint:disable-next-line:no-bitwise
            out += (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
        }
        return out;
    }
}
