// @gsp-svc/cloudprint/esm2015/lib/utils/cloudprint.Urls.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Config } from '../entitys/config';
/** @type {?} */
export const Urls = {
    // 第三方应用都统一走自己服务端转调
    /**
     * 云打印的基地址
     */
    CloudPrintBaseUrl: '/api/runtime/' + Config.PRINT_SU + '/v1.0/service',
    CloudPrintDimensionUrl: '/api/runtime/' + Config.PRINT_SU + '/v1.0/demensionservice',
    CloudPrintDeviceUrl: '/api/runtime/' + Config.PRINT_SU + '/v1.0/device',
    /**
     * 根据数据源获取格式
     */
    GetPrintFormatByDataSrouceId: '/GetPrintFormatByDataSrouceId?',
    /**
     * 根据数据源和维度获取格式
     */
    GetPrintFormatByDataSrouceAndDimension: '/GetPrintFormatByDataSrouceAndDimension?',
    /**
     * 获取格式和打印设备
     */
    GetPrintFormatAndDevice: '/getPrintFormatAndDevice?',
    GetPrintDevice: '/getPrintDevice?',
    /**
     * 保存格式
     */
    SavePrintFormat: '/SavePrintFormat',
    /**
     * 保存格式
     */
    SavePrintFormatWithRes: '/savePrintFormatWithRes',
    /**
     * 删除指定维度
     */
    DeletePrintFormat: '/DeletePrintFormat',
    /**
     * 获取指定业务对象下的打印数据源
     */
    GetDataSourceByBOId: '/GetDataSourceByBOId?',
    CreatePrintFormatWithBE: '/CreatePrintFormatWithBE?',
    CreatePrintFormatWithVO: '/CreatePrintFormatWithVO?',
    CreatePrintFormatWithBO: '/CreatePrintFormatWithBO?',
    CreatePrintFormatWithSchema: '/CreatePrintFormatWithSchema',
    /**
     * 加载已有格式API
     */
    LoadPrintFormatById: '/GetPrintFormatById?',
    LoadNewSourcePrintFormatById: '/getNewSourcePrintFormatById?',
    LoadNewSourcePrintFormatByIdAndLang: '/getNewSourcePrintFormatByIdAndLang?',
    /**
     * 加载打印数据API
     */
    LoadPrintData: '/GetPrintData?',
    /**
     * 获取设计器页面
     */
    GetDesigerHtml: '/GetDesigerHtml',
    /**
     * 获取扩展的js业务函数
     */
    GetCustomJsFunc: '/GetCustomJsFunc',
    /**
     * 打印API
     */
    PrintSingleById: '/OutputBEData?',
    PrintSingleByFilter: '/OutputBEDataWithFilter?',
    PrintMultiById: '/printMultiById?',
    PrintMultiByFilter: '/printMultiByFilter?',
    PrintSingleBOData: '/OutputBOData?',
    /**
     * 打印测试数据，用于设计时预览
     */
    PrintTestData: '/printTestData?',
    /**
     * 触发打印事件
     */
    SendPrintEvent: '/sendPrintEvent?',
    PrintAttachment: '/printAttachment?',
    IsRemotePrint: '/isRemotePrint?',
    GetEnableFonts: '/fonts?'
};
