// @gsp-sys/sysmgr-ui/esm2015/lib/func-op-help/func-op.help.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { SysFuncOperationComponent } from './sys-func-operation/sys-func-operation.component';
import { SysFuncOpHelpComponent } from './sys-func-op-help/sys-func-op-help.component';
import { BizObjectComponent } from './biz-object/biz-object.component';
import { BizObjectService } from './service/biz-object.service';
import { SysFuncOperationService } from './service/sys-func-operation.service';
import { SysFrontUtil } from '../sys-common/sys-front.util';
import { ModalModule } from '@farris/ui-modal';
import { MessagerModule } from '@farris/ui-messager';
import { DatagridModule } from '@farris/ui-datagrid';
import { TreeTableModule } from '@farris/ui-treetable';
import { LayoutModule } from '@farris/ui-layout';
import { SysFarrisFuncOperationComponent } from './sys-farris-func-operation/sys-farris-func-operation.component';
import { TreeViewModule } from '@progress/kendo-angular-treeview';
import { GridModule } from '@progress/kendo-angular-grid';
import { LayoutModule as KendoLayoutModule } from '@progress/kendo-angular-layout';
import { SysLocaleService } from '../locale/sys-locale.service';
export class SysFuncOpHelpModule {
}
SysFuncOpHelpModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    HttpClientModule,
                    FormsModule,
                    ReactiveFormsModule,
                    DatagridModule,
                    TreeTableModule,
                    LayoutModule,
                    TreeViewModule,
                    KendoLayoutModule,
                    LayoutModule,
                    GridModule,
                    ModalModule.forRoot(),
                    MessagerModule.forRoot()
                ],
                declarations: [
                    SysFuncOperationComponent,
                    SysFuncOpHelpComponent,
                    BizObjectComponent,
                    SysFarrisFuncOperationComponent
                ],
                providers: [
                    SysFrontUtil,
                    SysLocaleService,
                    SysFuncOperationService,
                    BizObjectService
                ],
                entryComponents: [
                    SysFuncOperationComponent,
                    SysFarrisFuncOperationComponent
                ],
                exports: [
                    SysFuncOperationComponent,
                    SysFuncOpHelpComponent,
                    BizObjectComponent,
                    SysFarrisFuncOperationComponent
                ]
            },] }
];
