// @gsp-svc/file-viewer/esm2015/lib/ofd-viewer/ofd-viewer/ofd-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
export class OfdViewerComponent {
    constructor() { }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.ocx = suwell.ofdReaderInit("ofdDiv", "100%", "100%");
        this.ocx.setCompsiteVisible("f_open", true);
        this.ocx.openFile(this.filePath, true);
    }
    /**
     * @param {?} filePath
     * @return {?}
     */
    fileChange(filePath) {
        this.filePath = filePath;
        this.ocx.openFile(filePath, true);
    }
}
OfdViewerComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-ofd-viewer',
                template: "<div id=\"ofdDiv\" style=\"width:100%;height:100%;\" ></div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
OfdViewerComponent.ctorParameters = () => [];
OfdViewerComponent.propDecorators = {
    filePath: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    OfdViewerComponent.prototype.filePath;
    /** @type {?} */
    OfdViewerComponent.prototype.ocx;
}
