// @edp-bif/runtime-api/esm2015/lib/bill-tracker/args/CanTrackingResult.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/args/CanTrackingResult.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from '@edp-aif/common-api';
import { CanTrackingInfo } from '../../entity/tracking/CanTrackingInfo';
export class CanTrackingResult extends CanTrackingInfo {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.billTrackingSettings = null;
        this.voName = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "billTrackingSettings", this.billTrackingSettings);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voName", this.voName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.billTrackingSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billTrackingSettings");
        this.voName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voName");
    }
}
if (false) {
    /**
     * 现为表组件展示字段，原为所有配置信息
     * @type {?}
     */
    CanTrackingResult.prototype.billTrackingSettings;
    /** @type {?} */
    CanTrackingResult.prototype.voName;
}
