// @edp-bif/runtime-api/esm2015/lib/bill-tracker/args/EntityDataResult.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/args/EntityDataResult.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper, BillSchema, BillData, } from '@edp-aif/common-api';
export class EntityDataResult extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.billSchema = null;
        this.billData = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "billSchema", this.billSchema);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billData", this.billData);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.billSchema = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "billSchema", BillSchema);
        this.billData = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "billData", BillData);
    }
}
if (false) {
    /** @type {?} */
    EntityDataResult.prototype.billSchema;
    /** @type {?} */
    EntityDataResult.prototype.billData;
}
