// @edp-bif/runtime-api/esm2015/lib/bill-tracker/args/GetChildrenEntityDataArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/args/GetChildrenEntityDataArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from '@edp-aif/common-api';
export class GetChildrenEntityDataArgs extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.voId = null;
        this.voMetadataId = null;
        this.suCode = null;
        this.nodeCodes = null;
        this.hierachyIdList = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voMetadataId", this.voMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "nodeCodes", BpmnModelJsonHelper.StringListToJson(this.nodeCodes));
        BpmnModelJsonHelper.SetAttributeValue(obj, "hierachyIdList", BpmnModelJsonHelper.StringListToJson(this.hierachyIdList));
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
        this.voMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voMetadataId");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.nodeCodes = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "nodeCodes"));
        this.hierachyIdList = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "hierachyIdList"));
    }
}
if (false) {
    /** @type {?} */
    GetChildrenEntityDataArgs.prototype.voId;
    /** @type {?} */
    GetChildrenEntityDataArgs.prototype.voMetadataId;
    /** @type {?} */
    GetChildrenEntityDataArgs.prototype.suCode;
    /**
     * @type {?}
     * @private
     */
    GetChildrenEntityDataArgs.prototype.nodeCodes;
    /**
     * @type {?}
     * @private
     */
    GetChildrenEntityDataArgs.prototype.hierachyIdList;
}
