// @edp-bif/runtime-api/esm2015/lib/bill-tracker/args/GetEntityDataArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/args/GetEntityDataArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { BillChildInfo } from '../../entity/tracking/BillChildInfo';
export class GetEntityDataArgs extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.billChildInfos = [];
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.schemaId = null;
        this.dataId = null;
        this.suCode = null;
        this.voMetadataId = null;
        this.billChildInfos = [];
        this.trackingSourceType = null;
        this.curYear = null;
        this.documentTypeId = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaId", this.schemaId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dataId", this.dataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voMetadataId", this.voMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billChildInfos", this.billChildInfos);
        BpmnModelJsonHelper.SetAttributeValue(obj, "trackingSourceType", this.trackingSourceType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "curYear", this.curYear);
        BpmnModelJsonHelper.SetAttributeValue(obj, "documentTypeId", this.documentTypeId);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.schemaId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaId");
        this.dataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dataId");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.voMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voMetadataId");
        this.billChildInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "billChildInfos", BillChildInfo);
        this.trackingSourceType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "trackingSourceType");
        this.curYear = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "curYear");
        this.documentTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "documentTypeId");
    }
}
if (false) {
    /** @type {?} */
    GetEntityDataArgs.prototype.schemaId;
    /** @type {?} */
    GetEntityDataArgs.prototype.dataId;
    /** @type {?} */
    GetEntityDataArgs.prototype.suCode;
    /** @type {?} */
    GetEntityDataArgs.prototype.voMetadataId;
    /** @type {?} */
    GetEntityDataArgs.prototype.billChildInfos;
    /** @type {?} */
    GetEntityDataArgs.prototype.trackingSourceType;
    /** @type {?} */
    GetEntityDataArgs.prototype.curYear;
    /** @type {?} */
    GetEntityDataArgs.prototype.documentTypeId;
}
