// @edp-bif/runtime-api/esm2015/lib/bill-tracker/args/GetJointQueryParametersArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/args/GetJointQueryParametersArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { JointQueryParametersContent } from "../../entity/tracking/JointQueryParametersContent";
export class GetJointQueryParametersArgs extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.contents = [];
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.trackingSourceType = null;
        this.suCode = null;
        this.schemaId = null;
        this.dataId = null;
        this.contents = [];
        this.curYear = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "trackingSourceType", this.trackingSourceType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaId", this.schemaId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dataId", this.dataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "contents", this.contents);
        BpmnModelJsonHelper.SetAttributeValue(obj, "curYear", this.curYear);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.trackingSourceType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "trackingSourceType");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.schemaId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaId");
        this.dataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dataId");
        this.contents = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "contents", JointQueryParametersContent);
        this.curYear = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "curYear");
    }
}
if (false) {
    /** @type {?} */
    GetJointQueryParametersArgs.prototype.trackingSourceType;
    /** @type {?} */
    GetJointQueryParametersArgs.prototype.suCode;
    /** @type {?} */
    GetJointQueryParametersArgs.prototype.schemaId;
    /** @type {?} */
    GetJointQueryParametersArgs.prototype.dataId;
    /** @type {?} */
    GetJointQueryParametersArgs.prototype.curYear;
    /** @type {?} */
    GetJointQueryParametersArgs.prototype.contents;
}
