// @edp-bif/runtime-api/esm5/lib/ref/mxgraph/Handler.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ref/mxgraph/Handler.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { mxgraph } from './mxgraphFactory';
/**
 * Handler
 * @type {?}
 */
var mxCellHighlightType = mxgraph.mxCellHighlight;
export { mxCellHighlightType as mxCellHighlight };
/** @type {?} */
var mxCellMarkerType = mxgraph.mxCellMarker;
export { mxCellMarkerType as mxCellMarker };
/** @type {?} */
var mxCellTrackerType = mxgraph.mxCellTracker;
export { mxCellTrackerType as mxCellTracker };
/** @type {?} */
var mxConnectionHandlerType = mxgraph.mxConnectionHandler;
export { mxConnectionHandlerType as mxConnectionHandler };
/** @type {?} */
var mxConstraintHandlerType = mxgraph.mxConstraintHandler;
export { mxConstraintHandlerType as mxConstraintHandler };
/** @type {?} */
var mxEdgeHandlerType = mxgraph.mxEdgeHandler;
export { mxEdgeHandlerType as mxEdgeHandler };
/** @type {?} */
var mxEdgeSegmentHandlerType = mxgraph.mxEdgeSegmentHandler;
export { mxEdgeSegmentHandlerType as mxEdgeSegmentHandler };
/** @type {?} */
var mxElbowEdgeHandlerType = mxgraph.mxElbowEdgeHandler;
export { mxElbowEdgeHandlerType as mxElbowEdgeHandler };
/** @type {?} */
var mxGraphHandlerType = mxgraph.mxGraphHandler;
export { mxGraphHandlerType as mxGraphHandler };
/** @type {?} */
var mxHandleType = mxgraph.mxHandle;
export { mxHandleType as mxHandle };
/** @type {?} */
var mxKeyHandlerType = mxgraph.mxKeyHandler;
export { mxKeyHandlerType as mxKeyHandler };
/** @type {?} */
var mxPanningHandlerType = mxgraph.mxPanningHandler;
export { mxPanningHandlerType as mxPanningHandler };
/** @type {?} */
var mxPopupMenuHandlerType = mxgraph.mxPopupMenuHandler;
export { mxPopupMenuHandlerType as mxPopupMenuHandler };
/** @type {?} */
var mxRubberbandType = mxgraph.mxRubberband;
export { mxRubberbandType as mxRubberband };
/** @type {?} */
var mxSelectionCellsHandlerType = mxgraph.mxSelectionCellsHandler;
export { mxSelectionCellsHandlerType as mxSelectionCellsHandler };
/** @type {?} */
var mxTooltipHandlerType = mxgraph.mxTooltipHandler;
export { mxTooltipHandlerType as mxTooltipHandler };
/** @type {?} */
var mxVertexHandlerType = mxgraph.mxVertexHandler;
export { mxVertexHandlerType as mxVertexHandler };
