// @edp-bif/runtime-api/esm5/lib/ref/mxgraph/Layout.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ref/mxgraph/Layout.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { mxgraph } from './mxgraphFactory';
/**
 * Layout
 * @type {?}
 */
var mxCircleLayoutType = mxgraph.mxCircleLayout;
export { mxCircleLayoutType as mxCircleLayout };
/** @type {?} */
var mxCompactTreeLayoutType = mxgraph.mxCompactTreeLayout;
export { mxCompactTreeLayoutType as mxCompactTreeLayout };
/** @type {?} */
var mxCompositeLayoutType = mxgraph.mxCompositeLayout;
export { mxCompositeLayoutType as mxCompositeLayout };
/** @type {?} */
var mxEdgeLabelLayoutType = mxgraph.mxEdgeLabelLayout;
export { mxEdgeLabelLayoutType as mxEdgeLabelLayout };
/** @type {?} */
var mxFastOrganicLayoutType = mxgraph.mxFastOrganicLayout;
export { mxFastOrganicLayoutType as mxFastOrganicLayout };
/** @type {?} */
var mxGraphLayoutType = mxgraph.mxGraphLayout;
export { mxGraphLayoutType as mxGraphLayout };
/** @type {?} */
var mxParallelEdgeLayoutType = mxgraph.mxParallelEdgeLayout;
export { mxParallelEdgeLayoutType as mxParallelEdgeLayout };
/** @type {?} */
var mxPartitionLayoutType = mxgraph.mxPartitionLayout;
export { mxPartitionLayoutType as mxPartitionLayout };
/** @type {?} */
var mxRadialTreeLayoutType = mxgraph.mxRadialTreeLayout;
export { mxRadialTreeLayoutType as mxRadialTreeLayout };
/** @type {?} */
var mxStackLayoutType = mxgraph.mxStackLayout;
export { mxStackLayoutType as mxStackLayout };
/**
 * Layout Hierarchical
 * @type {?}
 */
var mxHierarchicalLayoutType = mxgraph.mxHierarchicalLayout;
export { mxHierarchicalLayoutType as mxHierarchicalLayout };
/** @type {?} */
var mxSwimlaneLayoutType = mxgraph.mxSwimlaneLayout;
export { mxSwimlaneLayoutType as mxSwimlaneLayout };
/**
 * Layout Hierarchical Model
 * @type {?}
 */
var mxGraphAbstractHierarchyCellType = mxgraph.mxGraphAbstractHierarchyCell;
export { mxGraphAbstractHierarchyCellType as mxGraphAbstractHierarchyCell };
/** @type {?} */
var mxGraphHierarchyEdgeType = mxgraph.mxGraphHierarchyEdge;
export { mxGraphHierarchyEdgeType as mxGraphHierarchyEdge };
/** @type {?} */
var mxGraphHierarchyModelType = mxgraph.mxGraphHierarchyModel;
export { mxGraphHierarchyModelType as mxGraphHierarchyModel };
/** @type {?} */
var mxGraphHierarchyNodeType = mxgraph.mxGraphHierarchyNode;
export { mxGraphHierarchyNodeType as mxGraphHierarchyNode };
/** @type {?} */
var mxSwimlaneModelType = mxgraph.mxSwimlaneModel;
export { mxSwimlaneModelType as mxSwimlaneModel };
/**
 * Layout Hierarchical Stage
 * @type {?}
 */
var mxCoordinateAssignmentType = mxgraph.mxCoordinateAssignment;
export { mxCoordinateAssignmentType as mxCoordinateAssignment };
/** @type {?} */
var mxHierarchicalLayoutStageType = mxgraph.mxHierarchicalLayoutStage;
export { mxHierarchicalLayoutStageType as mxHierarchicalLayoutStage };
/** @type {?} */
var mxMedianHybridCrossingReductionType = mxgraph.mxMedianHybridCrossingReduction;
export { mxMedianHybridCrossingReductionType as mxMedianHybridCrossingReduction };
/** @type {?} */
var mxMinimumCycleRemoverType = mxgraph.mxMinimumCycleRemover;
export { mxMinimumCycleRemoverType as mxMinimumCycleRemover };
/** @type {?} */
var mxSwimlaneOrderingType = mxgraph.mxSwimlaneOrdering;
export { mxSwimlaneOrderingType as mxSwimlaneOrdering };
