// @edp-bif/runtime-api/esm5/lib/ref/mxgraph/Shape.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ref/mxgraph/Shape.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { mxgraph } from './mxgraphFactory';
/**
 * Shape
 * @type {?}
 */
var mxActorType = mxgraph.mxActor;
export { mxActorType as mxActor };
/** @type {?} */
var mxArrowType = mxgraph.mxArrow;
export { mxArrowType as mxArrow };
/** @type {?} */
var mxArrowConnectorType = mxgraph.mxArrowConnector;
export { mxArrowConnectorType as mxArrowConnector };
/** @type {?} */
var mxCloudType = mxgraph.mxCloud;
export { mxCloudType as mxCloud };
/** @type {?} */
var mxConnectorType = mxgraph.mxConnector;
export { mxConnectorType as mxConnector };
/** @type {?} */
var mxCylinderType = mxgraph.mxCylinder;
export { mxCylinderType as mxCylinder };
/** @type {?} */
var mxDoubleEllipseType = mxgraph.mxDoubleEllipse;
export { mxDoubleEllipseType as mxDoubleEllipse };
/** @type {?} */
var mxEllipseType = mxgraph.mxEllipse;
export { mxEllipseType as mxEllipse };
/** @type {?} */
var mxHexagonType = mxgraph.mxHexagon;
export { mxHexagonType as mxHexagon };
/** @type {?} */
var mxImageShapeType = mxgraph.mxImageShape;
export { mxImageShapeType as mxImageShape };
/** @type {?} */
var mxLabelType = mxgraph.mxLabel;
export { mxLabelType as mxLabel };
/** @type {?} */
var mxLineType = mxgraph.mxLine;
export { mxLineType as mxLine };
/** @type {?} */
var mxMarkerType = mxgraph.mxMarker;
export { mxMarkerType as mxMarker };
/** @type {?} */
var mxPolylineType = mxgraph.mxPolyline;
export { mxPolylineType as mxPolyline };
/** @type {?} */
var mxRectangleShapeType = mxgraph.mxRectangleShape;
export { mxRectangleShapeType as mxRectangleShape };
/** @type {?} */
var mxRhombusType = mxgraph.mxRhombus;
export { mxRhombusType as mxRhombus };
/** @type {?} */
var mxShapeType = mxgraph.mxShape;
export { mxShapeType as mxShape };
/** @type {?} */
var mxStencilType = mxgraph.mxStencil;
export { mxStencilType as mxStencil };
/** @type {?} */
var mxStencilRegistryType = mxgraph.mxStencilRegistry;
export { mxStencilRegistryType as mxStencilRegistry };
/** @type {?} */
var mxSwimlaneType = mxgraph.mxSwimlane;
export { mxSwimlaneType as mxSwimlane };
/** @type {?} */
var mxTextType = mxgraph.mxText;
export { mxTextType as mxText };
/** @type {?} */
var mxTriangleType = mxgraph.mxTriangle;
export { mxTriangleType as mxTriangle };
