// @edp-bif/runtime-api/esm5/lib/ref/mxgraph/Util.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ref/mxgraph/Util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { mxgraph } from './mxgraphFactory';
/**
 * Util
 * @type {?}
 */
var mxAbstractCanvas2DType = mxgraph.mxAbstractCanvas2D;
export { mxAbstractCanvas2DType as mxAbstractCanvas2D };
/** @type {?} */
var mxAnimationType = mxgraph.mxAnimation;
export { mxAnimationType as mxAnimation };
/** @type {?} */
var mxAutoSaveManagerType = mxgraph.mxAutoSaveManager;
export { mxAutoSaveManagerType as mxAutoSaveManager };
/** @type {?} */
var mxClipboardType = mxgraph.mxClipboard;
export { mxClipboardType as mxClipboard };
/** @type {?} */
var mxConstantsType = mxgraph.mxConstants;
export { mxConstantsType as mxConstants };
/** @type {?} */
var mxDictionaryType = mxgraph.mxDictionary;
export { mxDictionaryType as mxDictionary };
/** @type {?} */
var mxDivResizerType = mxgraph.mxDivResizer;
export { mxDivResizerType as mxDivResizer };
/** @type {?} */
var mxDragSourceType = mxgraph.mxDragSource;
export { mxDragSourceType as mxDragSource };
/** @type {?} */
var mxEffectsType = mxgraph.mxEffects;
export { mxEffectsType as mxEffects };
/** @type {?} */
var mxEventType = mxgraph.mxEvent;
export { mxEventType as mxEvent };
/** @type {?} */
var mxEventObjectType = mxgraph.mxEventObject;
export { mxEventObjectType as mxEventObject };
/** @type {?} */
var mxEventSourceType = mxgraph.mxEventSource;
export { mxEventSourceType as mxEventSource };
/** @type {?} */
var mxFormType = mxgraph.mxForm;
export { mxFormType as mxForm };
/** @type {?} */
var mxGuideType = mxgraph.mxGuide;
export { mxGuideType as mxGuide };
/** @type {?} */
var mxImageType = mxgraph.mxImage;
export { mxImageType as mxImage };
/** @type {?} */
var mxImageBundleType = mxgraph.mxImageBundle;
export { mxImageBundleType as mxImageBundle };
/** @type {?} */
var mxImageExportType = mxgraph.mxImageExport;
export { mxImageExportType as mxImageExport };
/** @type {?} */
var mxLogType = mxgraph.mxLog;
export { mxLogType as mxLog };
/** @type {?} */
var mxMorphingType = mxgraph.mxMorphing;
export { mxMorphingType as mxMorphing };
/** @type {?} */
var mxMouseEventType = mxgraph.mxMouseEvent;
export { mxMouseEventType as mxMouseEvent };
/** @type {?} */
var mxObjectIdentityType = mxgraph.mxObjectIdentity;
export { mxObjectIdentityType as mxObjectIdentity };
/** @type {?} */
var mxPanningManagerType = mxgraph.mxPanningManager;
export { mxPanningManagerType as mxPanningManager };
/** @type {?} */
var mxPointType = mxgraph.mxPoint;
export { mxPointType as mxPoint };
/** @type {?} */
var mxPopupMenuType = mxgraph.mxPopupMenu;
export { mxPopupMenuType as mxPopupMenu };
/** @type {?} */
var mxRectangleType = mxgraph.mxRectangle;
export { mxRectangleType as mxRectangle };
/** @type {?} */
var mxResourcesType = mxgraph.mxResources;
export { mxResourcesType as mxResources };
/** @type {?} */
var mxSvgCanvas2DType = mxgraph.mxSvgCanvas2D;
export { mxSvgCanvas2DType as mxSvgCanvas2D };
/** @type {?} */
var mxToolbarType = mxgraph.mxToolbar;
export { mxToolbarType as mxToolbar };
/** @type {?} */
var mxUndoableEditType = mxgraph.mxUndoableEdit;
export { mxUndoableEditType as mxUndoableEdit };
/** @type {?} */
var mxUndoManagerType = mxgraph.mxUndoManager;
export { mxUndoManagerType as mxUndoManager };
/** @type {?} */
var mxUrlConverterType = mxgraph.mxUrlConverter;
export { mxUrlConverterType as mxUrlConverter };
/** @type {?} */
var mxUtilsType = mxgraph.mxUtils;
export { mxUtilsType as mxUtils };
/** @type {?} */
var mxVmlCanvas2DType = mxgraph.mxVmlCanvas2D;
export { mxVmlCanvas2DType as mxVmlCanvas2D };
/** @type {?} */
var mxWindowType = mxgraph.mxWindow;
export { mxWindowType as mxWindow };
/** @type {?} */
var mxXmlCanvas2DType = mxgraph.mxXmlCanvas2D;
export { mxXmlCanvas2DType as mxXmlCanvas2D };
/** @type {?} */
var mxXmlRequestType = mxgraph.mxXmlRequest;
export { mxXmlRequestType as mxXmlRequest };
