// @edp-bif/runtime-api/esm5/lib/ref/mxgraph/View.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ref/mxgraph/View.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { mxgraph } from './mxgraphFactory';
/**
 * View
 * @type {?}
 */
var mxCellEditorType = mxgraph.mxCellEditor;
export { mxCellEditorType as mxCellEditor };
/** @type {?} */
var mxCellOverlayType = mxgraph.mxCellOverlay;
export { mxCellOverlayType as mxCellOverlay };
/** @type {?} */
var mxCellRendererType = mxgraph.mxCellRenderer;
export { mxCellRendererType as mxCellRenderer };
/** @type {?} */
var mxCellStateType = mxgraph.mxCellState;
export { mxCellStateType as mxCellState };
/** @type {?} */
var mxCellStatePreviewType = mxgraph.mxCellStatePreview;
export { mxCellStatePreviewType as mxCellStatePreview };
/** @type {?} */
var mxConnectionConstraintType = mxgraph.mxConnectionConstraint;
export { mxConnectionConstraintType as mxConnectionConstraint };
/** @type {?} */
var mxEdgeStyleType = mxgraph.mxEdgeStyle;
export { mxEdgeStyleType as mxEdgeStyle };
/** @type {?} */
var mxGraphType = mxgraph.mxGraph;
export { mxGraphType as mxGraph };
/** @type {?} */
var mxGraphSelectionModelType = mxgraph.mxGraphSelectionModel;
export { mxGraphSelectionModelType as mxGraphSelectionModel };
/** @type {?} */
var mxGraphViewType = mxgraph.mxGraphView;
export { mxGraphViewType as mxGraphView };
/** @type {?} */
var mxLayoutManagerType = mxgraph.mxLayoutManager;
export { mxLayoutManagerType as mxLayoutManager };
/** @type {?} */
var mxMultiplicityType = mxgraph.mxMultiplicity;
export { mxMultiplicityType as mxMultiplicity };
/** @type {?} */
var mxOutlineType = mxgraph.mxOutline;
export { mxOutlineType as mxOutline };
/** @type {?} */
var mxPerimeterType = mxgraph.mxPerimeter;
export { mxPerimeterType as mxPerimeter };
/** @type {?} */
var mxPrintPreviewType = mxgraph.mxPrintPreview;
export { mxPrintPreviewType as mxPrintPreview };
/** @type {?} */
var mxStyleRegistryType = mxgraph.mxStyleRegistry;
export { mxStyleRegistryType as mxStyleRegistry };
/** @type {?} */
var mxStylesheetType = mxgraph.mxStylesheet;
export { mxStylesheetType as mxStylesheet };
/** @type {?} */
var mxSwimlaneManagerType = mxgraph.mxSwimlaneManager;
export { mxSwimlaneManagerType as mxSwimlaneManager };
/** @type {?} */
var mxTemporaryCellStatesType = mxgraph.mxTemporaryCellStates;
export { mxTemporaryCellStatesType as mxTemporaryCellStates };
