// @farris/ui-datepicker/esm5/lib/components/selection-sidebar/selection-sidebar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-02 14:14:22
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-09-04 15:03:00
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
var SelectionSidebarComponent = /** @class */ (function () {
    function SelectionSidebarComponent() {
        this.setTime = new EventEmitter();
    }
    SelectionSidebarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-selection-sidebar',
                    template: "<div class=\"f-datepicker-sidebar\">\r\n    <button *ngFor=\"let item of opts.shortcuts\" (click)=\"item.onClick(setTime)\">\r\n        {{ item.text }}\r\n    </button>\r\n</div>\r\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    SelectionSidebarComponent.propDecorators = {
        opts: [{ type: Input }],
        setTime: [{ type: Output }]
    };
    return SelectionSidebarComponent;
}());
export { SelectionSidebarComponent };
if (false) {
    /** @type {?} */
    SelectionSidebarComponent.prototype.opts;
    /** @type {?} */
    SelectionSidebarComponent.prototype.setTime;
}
