// @farris/ui-forms/esm5/lib/forms.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { IdService, FarrisCommonModule } from '@farris/ui-common';
import { RadioComponent } from './radio/radio.component';
import { RadioButtonComponent } from './radio/radio-button.component';
import { RadioGroupComponent } from './radio/radio-group.component';
import { FarrisRadioGroupComponent } from './radio/radiogroup.component';
import { CheckboxComponent } from './checkbox/checkbox.component';
import { CheckboxGroupComponent } from './checkbox/checkbox-group.component';
import { FarrisCheckboxGroupComponent } from './checkbox/checkboxgroup.component';
import { TextareaAutoSizeDirective } from './textarea-autosize.directive';
import { TextareaZoomDirective } from './textarea-zoom.directive';
import { InputAppendDirective } from './input-append.directive';
import { MessagerModule } from '@farris/ui-messager';
import { ShortCutTabKeyDirective } from './shortcut-tab.directive';
import { TextareaWordcountDirective } from './textarea-wordcount.directive';
import { LocaleModule } from '@farris/ui-locale';
import { TextareaCommentsDirective } from './textarea-comments.directive';
import { SingleListComponent } from './comments/single-list.component';
import { SingleSelectDirective } from './single-select.directive';
var FarrisFormsModule = /** @class */ (function () {
    function FarrisFormsModule() {
    }
    FarrisFormsModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        RadioComponent,
                        RadioGroupComponent,
                        CheckboxComponent,
                        CheckboxGroupComponent,
                        RadioButtonComponent,
                        FarrisRadioGroupComponent,
                        FarrisCheckboxGroupComponent,
                        TextareaAutoSizeDirective,
                        TextareaZoomDirective,
                        ShortCutTabKeyDirective,
                        TextareaWordcountDirective,
                        InputAppendDirective,
                        TextareaCommentsDirective,
                        SingleListComponent,
                        SingleSelectDirective
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        MessagerModule.forRoot(),
                        FarrisCommonModule.forRoot(),
                        LocaleModule,
                    ],
                    exports: [
                        RadioComponent,
                        RadioGroupComponent,
                        CheckboxComponent,
                        CheckboxGroupComponent,
                        RadioButtonComponent,
                        FarrisRadioGroupComponent,
                        FarrisCheckboxGroupComponent,
                        TextareaAutoSizeDirective,
                        TextareaZoomDirective,
                        ShortCutTabKeyDirective,
                        TextareaWordcountDirective,
                        InputAppendDirective,
                        TextareaCommentsDirective,
                        SingleSelectDirective
                    ],
                    entryComponents: [
                        SingleListComponent
                    ],
                    providers: [IdService],
                },] }
    ];
    return FarrisFormsModule;
}());
export { FarrisFormsModule };
