// @farris/ui-forms/esm5/lib/shortcut-tab.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventManager } from '@angular/platform-browser';
import { Directive, ElementRef, Injector } from '@angular/core';
var ShortCutTabKeyDirective = /** @class */ (function () {
    function ShortCutTabKeyDirective(el, injector, eventManager) {
        this.el = el;
        this.injector = injector;
        this.eventManager = eventManager;
    }
    /**
     * @return {?}
     */
    ShortCutTabKeyDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    ShortCutTabKeyDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.eventManager.addEventListener(this.el.nativeElement, 'keydown', (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            e.stopPropagation();
            if (e.key === 'Tab') {
                _this.onTabKeydownHandler(e.target);
            }
        }));
    };
    /**
     * @private
     * @param {?} currentTarget
     * @return {?}
     */
    ShortCutTabKeyDirective.prototype.onTabKeydownHandler = /**
     * @private
     * @param {?} currentTarget
     * @return {?}
     */
    function (currentTarget) {
        /** @type {?} */
        var allInputs = this.el.nativeElement.querySelectorAll('input,textarea');
        allInputs = Array.from(allInputs);
        /** @type {?} */
        var currentInputIndex = allInputs.findIndex((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n === currentTarget; }));
        /** @type {?} */
        var nextInputIndex = currentInputIndex + 1;
        if (nextInputIndex >= allInputs.length) {
            nextInputIndex = 0;
        }
        /** @type {?} */
        var nextInputTarget = allInputs[nextInputIndex];
        if (nextInputTarget) {
            setTimeout((/**
             * @return {?}
             */
            function () {
                nextInputTarget.focus();
            }));
        }
    };
    ShortCutTabKeyDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'shortcut-tab',
                },] }
    ];
    /** @nocollapse */
    ShortCutTabKeyDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Injector },
        { type: EventManager }
    ]; };
    return ShortCutTabKeyDirective;
}());
export { ShortCutTabKeyDirective };
if (false) {
    /** @type {?} */
    ShortCutTabKeyDirective.prototype.el;
    /** @type {?} */
    ShortCutTabKeyDirective.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    ShortCutTabKeyDirective.prototype.eventManager;
}
