// @gsp-svc/expression/esm2015/lib/my-monaco-editor/my-monaco-editor.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MyMonacoEditorComponent } from './my-monaco-editor.component';
import { MY_MONACO_EDITOR_CONFIG } from './editorconfig';
import { CodeEditorEventService } from './my-monaco-editor-services/monaco-editor.event.service';
export class MyMonacoEditorModule {
    /**
     * @param {?=} config
     * @return {?}
     */
    static forRoot(config = {}) {
        return {
            ngModule: MyMonacoEditorModule,
            providers: [
                { provide: MY_MONACO_EDITOR_CONFIG, useValue: config }
            ]
        };
    }
}
MyMonacoEditorModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                exports: [
                    MyMonacoEditorComponent
                ],
                entryComponents: [
                    MyMonacoEditorComponent
                ],
                declarations: [
                    MyMonacoEditorComponent
                ],
                providers: [
                    CodeEditorEventService
                ]
            },] }
];
