// @gsp-svc/expression/esm2015/lib/expression-entity-info-grid/expression-entity-info-grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { Property } from '../schema';
import { CurrentNode } from '../currentNode';
import { ExpressionEditorLocalePipe } from '../expression-editor-locale.pipe';
import { NewProperty } from '../i18n/NewSchema';
export class ExpressionEntityInfoGridComponent {
    constructor() {
        this.schema = new Map();
        this.newSchemaMap = new Map(); //统一Schema
        //统一Schema
        this.choseEntityTableName = new CurrentNode;
        this.selectSchemaRow = new EventEmitter();
        this.dbclickInsert = new EventEmitter();
        this.selectString = "";
        this.items = [];
        this.columns = [];
        //初始化i18n标识
        this.localeId = localStorage.getItem('languageCode');
        this.localePipe = new ExpressionEditorLocalePipe(this.localeId);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngOnChanges() {
        // this.columns = [{ field: 'name', title: this.localePipe.transform("EntityPropList"), filter: true }];
        /** @type {?} */
        let schemaName = this.choseEntityTableName.schema;
        /** @type {?} */
        let parentName = this.choseEntityTableName.parentName;
        /** @type {?} */
        let tableId = this.choseEntityTableName.id;
        /** @type {?} */
        let tableName = this.choseEntityTableName.name;
        /** @type {?} */
        let schemaItem;
        if (this.newSchemaMap && this.newSchemaMap.size > 0) {
            //统一Schema
            this.columns = [{ field: 'code', title: this.localePipe.transform("EntityPropList"), filter: true }];
            this.items = [];
            schemaItem = this.newSchemaMap.get(schemaName);
            if (schemaItem) {
                /** @type {?} */
                let entityType = schemaItem.entityTypes.find((/**
                 * @param {?} entitytype
                 * @return {?}
                 */
                entitytype => (entitytype.code == tableName) || (entitytype.code == "")));
                if (!entityType) {
                    /** @type {?} */
                    let entitySet = schemaItem.entitySets.find((/**
                     * @param {?} entityset
                     * @return {?}
                     */
                    entityset => entityset.code == tableName));
                    if (entitySet) {
                        entityType = schemaItem.entityTypes.find((/**
                         * @param {?} entitytype
                         * @return {?}
                         */
                        entitytype => entitytype.code == entitySet.entityType));
                    }
                }
                this.items = [...entityType.properties];
                entityType.navigationProperties.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => {
                    /** @type {?} */
                    let property = new NewProperty;
                    property.code = item.code;
                    property.name = "";
                    property.dataType = item.navigationType.toString();
                    property.nullable = false;
                    property.defaultValue = null;
                    this.items.push(property);
                    this.items = Array.from(this.items);
                }));
                this.selectString = parentName + "." + entityType.name;
            }
        }
        else {
            //旧的Schema
            this.columns = [{ field: 'name', title: this.localePipe.transform("EntityPropList"), filter: true }];
            this.items = [];
            schemaItem = this.schema.get(schemaName);
            if (schemaItem) {
                /** @type {?} */
                let entityType = schemaItem.entityTypes.find((/**
                 * @param {?} entitytype
                 * @return {?}
                 */
                entitytype => (entitytype.id == tableId && entitytype.name == tableName) || (entitytype.id == "" && entitytype.name == tableName)));
                if (entityType) {
                    this.items = [...entityType.properties];
                    entityType.navigationProperties.forEach((/**
                     * @param {?} item
                     * @return {?}
                     */
                    item => {
                        /** @type {?} */
                        let property = new Property;
                        property.name = item.name;
                        property.type = item.type.toString();
                        property.description = item.description;
                        this.items.push(property);
                    }));
                    this.selectString = parentName + "." + entityType.name;
                }
                else {
                    /** @type {?} */
                    let entitySet = schemaItem.entitySets.find((/**
                     * @param {?} entityset
                     * @return {?}
                     */
                    entityset => entityset.name == tableName));
                    if (entitySet) {
                        /** @type {?} */
                        let eType = schemaItem.entityTypes.find((/**
                         * @param {?} entitytype
                         * @return {?}
                         */
                        entitytype => entitytype.name == entitySet.entityType));
                        this.items = [...eType.properties];
                        eType.navigationProperties.forEach((/**
                         * @param {?} item
                         * @return {?}
                         */
                        item => {
                            /** @type {?} */
                            let property = new Property;
                            property.name = item.name;
                            property.type = item.type.toString();
                            property.description = item.description;
                            this.items.push(property);
                        }));
                        this.selectString = parentName + "." + entitySet.name + ".[0]";
                    }
                }
            }
        }
    }
    //数据：子组件双击插入              (farris双击事件，调整)
    /**
     * @param {?} $event
     * @return {?}
     */
    ondblclickRow($event) {
        //this.dbclickInsert.emit(this.selectString+"."+$event.rowData.name);
        if (this.newSchemaMap && this.newSchemaMap.size > 0) {
            this.dbclickInsert.emit($event.rowData.data.code);
        }
        else if (this.schema && this.schema.size > 0) {
            this.dbclickInsert.emit($event.rowData.data.name);
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    selectRowChange($event) {
        this.selectSchemaRow.emit($event.data);
    }
}
ExpressionEntityInfoGridComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-expression-entity-info-grid',
                template: "<farris-datagrid\r\n[idField]=\"'name'\"\r\n[height]=\"268\"\r\n[columns]=\"columns\" \r\n[data]=\"items\"\r\n[fit]=\"true\"\r\n[fitColumns]=\"true\"\r\n[striped]=\"false\"\r\n[showHeader]=\"true\"\r\n[pagination]=\"false\"\r\n[enableFilterRow]=\"true\"\r\n(dblclickRow)=\"ondblclickRow($event)\"\r\n(selectChanged)=\"selectRowChange($event)\"\r\n>\r\n</farris-datagrid>",
                styles: [""]
            }] }
];
/** @nocollapse */
ExpressionEntityInfoGridComponent.ctorParameters = () => [];
ExpressionEntityInfoGridComponent.propDecorators = {
    schema: [{ type: Input }],
    newSchemaMap: [{ type: Input }],
    choseEntityTableName: [{ type: Input }],
    selectSchemaRow: [{ type: Output }],
    dbclickInsert: [{ type: Output }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ExpressionEntityInfoGridComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    ExpressionEntityInfoGridComponent.prototype.localePipe;
    /** @type {?} */
    ExpressionEntityInfoGridComponent.prototype.schema;
    /** @type {?} */
    ExpressionEntityInfoGridComponent.prototype.newSchemaMap;
    /** @type {?} */
    ExpressionEntityInfoGridComponent.prototype.choseEntityTableName;
    /** @type {?} */
    ExpressionEntityInfoGridComponent.prototype.selectSchemaRow;
    /** @type {?} */
    ExpressionEntityInfoGridComponent.prototype.dbclickInsert;
    /** @type {?} */
    ExpressionEntityInfoGridComponent.prototype.selectString;
    /** @type {?} */
    ExpressionEntityInfoGridComponent.prototype.items;
    /** @type {?} */
    ExpressionEntityInfoGridComponent.prototype.columns;
}
