// @gsp-cmp/ubcc-param-data-ui/esm2015/lib/ubcc-param-data-ui-dialog/ubcc-param-data-ui-dialog.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ubcc-param-data-ui-dialog/ubcc-param-data-ui-dialog.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input, TemplateRef, ViewChild } from '@angular/core';
import { UbccParamTemplate, UbccParamTemplateSerializer } from '@gsp-cmp/ubcc-param-vo';
import { UbccParamDataManangerService } from '../service/http-service';
import { DataTypeConst } from '@ecp-caf/common-structure';
import { UbccParamDataUtils } from '../utils/param-data-utils';
export class UbccParamDataUiDialogComponent {
    /**
     * @param {?} utils
     * @param {?} http
     */
    constructor(utils, http) {
        this.utils = utils;
        this.http = http;
        this.paramDataItems = [];
        this.confirmEmitter = new EventEmitter();
        this.closeEmitter = new EventEmitter();
        this.close = (/**
         * @return {?}
         */
        () => {
            if (this.dialog) {
                this.dialog.close();
            }
            this.closeEmitter.emit();
        });
    }
    /**
     * 输入数据为JSON对象
     * @param {?} data
     * @return {?}
     */
    set data(data) {
        this._data = data;
        this.loadParamData(this.templateId);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.template = this.template || new UbccParamTemplate();
    }
    /**
     * @private
     * @param {?} templateId
     * @return {?}
     */
    loadParamData(templateId) {
        if (!templateId) {
            return;
        }
        this.http.getParamTemplate(templateId).subscribe((/**
         * @param {?} vo
         * @return {?}
         */
        (vo) => {
            if (!vo.content) {
                return;
            }
            /** @type {?} */
            let serializer = new UbccParamTemplateSerializer();
            this.template = serializer.deserializer(JSON.parse(vo.content));
            this.getParamDataItemsStructure();
            this.convertJsonToParamData();
        }));
    }
    /**
     * @private
     * @return {?}
     */
    convertJsonToParamData() {
        if (!this._data || Object.keys(this._data).length <= 0) {
            return;
        }
        this.paramDataItems.map((/**
         * @param {?} item
         * @return {?}
         */
        (item) => {
            /** @type {?} */
            let key = Object.keys(this._data).find((/**
             * @param {?} x
             * @return {?}
             */
            x => x === item.code));
            if (!key || !this._data[key]) {
                return;
            }
            if (this.isCollection(item)) {
                //参数的值如果是数组，以逗号隔开
                item.value = this._data[key].join(',');
                if (!this.utils.isHelpType(item.parameterType)) {
                    return;
                }
                /** @type {?} */
                let helpId = ((/** @type {?} */ (this.utils.getCollectionElementType(item.parameterType)))).refId;
                this.http.getHelpData(helpId, this._data[key]).subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                (result) => {
                    if (!result || result.length <= 0) {
                        return;
                    }
                    /** @type {?} */
                    let values = [];
                    result.map((/**
                     * @param {?} resItem
                     * @return {?}
                     */
                    resItem => values.push(resItem.text)));
                    item.valueText = values.join(',');
                }));
            }
            else {
                item.value = this._data[key];
                if (item.parameterTypeDesc !== 'help') {
                    return;
                }
                /** @type {?} */
                let helpId = ((/** @type {?} */ (item.parameterType))).refId;
                this.http.getHelpData(helpId, [item.value]).subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                (result) => {
                    if (!result || result.length <= 0) {
                        return;
                    }
                    item.valueText = result[0].text;
                }));
            }
        }));
    }
    /**
     * @private
     * @return {?}
     */
    getParamDataItemsStructure() {
        if (!this.template || !this.template.params) {
            return;
        }
        this.paramDataItems = [];
        this.template.params.map((/**
         * @param {?} x
         * @return {?}
         */
        x => this.paramDataItems.push(x)));
    }
    /**
     * @private
     * @return {?}
     */
    covertParamDataToJson() {
        if (!this.paramDataItems || this.paramDataItems.length <= 0) {
            return null;
        }
        /** @type {?} */
        let object = new Object();
        this.paramDataItems.map((/**
         * @param {?} item
         * @return {?}
         */
        (item) => {
            if (this.isCollection(item)) {
                object[item.code] = [];
                /** @type {?} */
                let itemValues = item.value.split(',');
                if (!itemValues || itemValues.length <= 0) {
                    return;
                }
                itemValues.map((/**
                 * @param {?} itemValue
                 * @return {?}
                 */
                itemValue => object[item.code].push(itemValue)));
                //帮助name赋值
                if (item.parameterTypeDesc === 'help') {
                    object[`${item.code}_name`] = [];
                    /** @type {?} */
                    let itemTextValues = item.valueText.split(',');
                    if (!itemTextValues || itemTextValues.length <= 0) {
                        return;
                    }
                    itemTextValues.map((/**
                     * @param {?} itemTextValue
                     * @return {?}
                     */
                    itemTextValue => object[`${item.code}_name`].push(itemTextValue)));
                }
            }
            else {
                object[item.code] = item.value;
                //帮助name赋值
                if (item.parameterTypeDesc === 'help') {
                    object[`${item.code}_name`] = item.valueText;
                }
            }
        }));
        return object;
    }
    /**
     * @return {?}
     */
    confirm() {
        /** @type {?} */
        let jsonData = this.covertParamDataToJson();
        this.confirmEmitter.emit(jsonData);
    }
    /**
     * @private
     * @param {?} param
     * @return {?}
     */
    isCollection(param) {
        if (!param || !param.parameterType) {
            return false;
        }
        if (param.parameterType.getKind() === DataTypeConst.DEFAULT_COLLECTION) {
            return true;
        }
        else {
            return false;
        }
    }
}
UbccParamDataUiDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-ubcc-param-data-ui-dialog',
                template: "<div class=\"d-flex\" style=\"position: absolute;top: 0;bottom: 0;left: 0;right: 0;padding: 0px 8px 0px 8px;\">\n  <lib-ubcc-param-data-ui class=\"f-utils-fill\" [(template)]=\"template\" [(paramDataItems)]=\"paramDataItems\"  style=\"overflow-y: auto;\"></lib-ubcc-param-data-ui>\n</div>\n\n<ng-template #btnRef>\n  <div style=\"width: 100%;line-height: 50px; text-align: center;min-width: 140px;\">\n    <button class=\"btn btn-primary\" (click)=confirm() style=\"float:right;margin-right: 5px\">\u786E\u5B9A</button>\n    <button class=\"btn btn-secondary\" (click)=close() style=\"float:right;margin-right: 5px\">\u5173\u95ED</button>\n  </div>\n</ng-template>",
                styles: [":host{display:flex;flex-direction:column;flex:1}"]
            }] }
];
/** @nocollapse */
UbccParamDataUiDialogComponent.ctorParameters = () => [
    { type: UbccParamDataUtils },
    { type: UbccParamDataManangerService }
];
UbccParamDataUiDialogComponent.propDecorators = {
    templateId: [{ type: Input }],
    data: [{ type: Input }],
    confirmEmitter: [{ type: Output }],
    closeEmitter: [{ type: Output }],
    btnRef: [{ type: ViewChild, args: ['btnRef',] }]
};
if (false) {
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.template;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype._data;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.templateId;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.paramDataItems;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.confirmEmitter;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.closeEmitter;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.btnRef;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.dialog;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.close;
    /**
     * @type {?}
     * @private
     */
    UbccParamDataUiDialogComponent.prototype.utils;
    /**
     * @type {?}
     * @private
     */
    UbccParamDataUiDialogComponent.prototype.http;
}
