// @farris/mobile-devkit/esm2015/lib/app/app_context.js
import { Context } from '../context/index';
import { BindingDataFactory } from '../binding-data';
import { VALIDATION_RULE_TOKEN } from './tokens';
class AppContext extends Context {
    /**
     * 构造函数
     */
    constructor(injector, eventBus, repositoryManager, bindingDataManager, viewModelContextManager, validationManager = {}, validationRule) {
        super();
        this.injector = injector;
        this.eventBus = eventBus;
        this.repositoryManager = repositoryManager;
        this.bindingDataManager = bindingDataManager;
        this.viewModelContextManager = viewModelContextManager;
        this.validationManager = validationManager;
        this.validationRule = validationRule;
        this.validationRule = injector.get(VALIDATION_RULE_TOKEN, []);
    }
    /**
     * 注册FrameContext
     */
    regViewModelContext(viewModelContext) {
        const repository = viewModelContext.repository;
        const repositoryName = repository.name;
        // Repository
        if (this.repositoryManager.ifRepositoryExits(repositoryName) === false) {
            this.repositoryManager.regRepository(repositoryName, repository);
        }
        // BindingData
        if (this.bindingDataManager.ifBindingDataExits(repositoryName) === false) {
            const bindingData = BindingDataFactory.createFromRepository(repository, '/');
            this.bindingDataManager.regBindingData(repositoryName, bindingData);
        }
        // 考虑路由再次进入的时候，AppContext没有被注销，但Component被再次构造的场景
        this.viewModelContextManager.regContext(viewModelContext);
    }
}
export { AppContext };
