// @farris/mobile-devkit/esm2015/lib/app/app_eventBus.js
import { Subject } from 'rxjs';
class AppEventBus {
    /**
     * 构造函数
     */
    constructor() {
        this.eventBus = new Subject();
        this.subscriptionsMap = new Map();
    }
    triggerEvent(event) {
        this.eventBus.next(event);
    }
    subscribe(componentId, func, funcError) {
        if (this.subscriptionsMap.has(componentId) === false) {
            const subscription = this.eventBus.subscribe((value) => func(value), (value) => funcError(value));
            this.subscriptionsMap.set(componentId, subscription);
        }
    }
    unsubscribe(componentId) {
        if (this.subscriptionsMap.has(componentId) === true) {
            this.subscriptionsMap.get(componentId).unsubscribe();
            this.subscriptionsMap.delete(componentId);
        }
    }
    /**
     * 清除注册得指定key名监听器
     * @param keyStr 监听器key 默认值为listview上注册得
     * @returns
     */
    unsubscribeAllListview(keyStr = '-listview-extend') {
        if (this.subscriptionsMap.size == 0) {
            return;
        }
        // @ts-ignore: Unreachable code error
        for (let key of this.subscriptionsMap.keys()) {
            if (key.indexOf(keyStr) > 0) {
                this.subscriptionsMap.get(key).unsubscribe();
                this.subscriptionsMap.delete(key);
            }
        }
    }
}
export { AppEventBus };
