// @farris/mobile-devkit/esm2015/lib/app/view_model_contex_manager.js
/**
 * ViewModelContext管理类
 */
class ViewModelContextManager {
    /**
     * 构造函数
     */
    constructor() {
        this.contextMap = new Map();
        this.contextSet = new Set();
    }
    /**
     * 注册Context
     */
    regContext(viewModelContext) {
        const id = viewModelContext.id;
        if (this.contextMap.has(id) === true) {
            // throw Error(`id为${id}的ViewModelContext已经存在`);
            // 临时处理，方式报错。
            this.unregContext(viewModelContext);
        }
        this.contextMap.set(id, viewModelContext);
        this.contextSet.add(viewModelContext);
    }
    /**
     * 取消注册
     */
    unregContext(context) {
        const id = context.id;
        this.contextMap.delete(id);
        this.contextSet.delete(context);
    }
    /**
     * 获取ContextMap
     */
    getContextMap() {
        return this.contextMap;
    }
    /**
     * 获取全部Context
     */
    getContexts() {
        return Array.from(this.contextSet);
    }
    /**
     * 根据命名控件获取上下文集合
     */
    getContextsByNamespace(namespace) {
        return this.getContexts();
    }
    /**
     * 根据id获取Context
     */
    getContextById(id) {
        const targetContext = this.contextMap.get(id);
        return targetContext;
    }
    /**
     * 获取根Context
     */
    getRootContext() {
        const contexts = this.getContexts();
        const rootContext = contexts.find((context) => {
            return context.parent === null;
        });
        return rootContext;
    }
    /**
     * 获取传入视图模型id的root及root的下一代集合数组
     */
    getRootContextAndPosterityById(viewModelId) {
        const targetContext = this.getContextById(viewModelId);
        const contexts = this.getContexts();
        let contextsGroup = [];
        // 1.找到ROOT
        const RootId = this.getContextsGroupRoot(targetContext);
        // 2.可直接遍历拿到所有直系后代 (目前只有两层结构直接获取下一代)
        contexts.map((context) => {
            if (context.parent && context.parent.id === RootId) {
                contextsGroup.push(context);
            }
        });
        contextsGroup.push(this.getContextById(RootId));
        return contextsGroup;
    }
    getContextsGroupRoot(context) {
        if (context.parent) {
            return this.getContextsGroupRoot(context.parent);
        }
        else {
            return context.id;
        }
    }
}
export { ViewModelContextManager };
