// @gsp-lcm/metadata-selector/esm5/metadata/metadata-select/components/metadatalist/metadatalistdetails/metadatalistdetails.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/metadatalist/metadatalistdetails/metadatalistdetails.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output } from '@angular/core';
var MetadatalistdetailsComponent = /** @class */ (function () {
    function MetadatalistdetailsComponent() {
        this.selectableSetting = { mode: 'single', checkboxOnly: false };
        this.dotnet = 'dotnet';
        this.java = 'java';
        this.detailSkip = 0;
        this.pageSize = 5;
        this.selectedPkgChanged = new EventEmitter();
        this.dataItem = [];
    }
    /**
     * @return {?}
     */
    MetadatalistdetailsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.mdItem && this.mdItem.repoPkg) {
            this.dataItem = this.mdItem.repoPkg;
            this.detailView = {
                data: this.dataItem.slice(this.detailSkip, this.detailSkip + this.pageSize),
                total: this.dataItem.length
            };
        }
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    MetadatalistdetailsComponent.prototype.detailPageChange = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var skip = _a.skip, take = _a.take;
        this.detailSkip = skip;
        this.detailView = {
            data: this.dataItem.slice(this.detailSkip, this.detailSkip + take),
            total: this.dataItem.length
        };
    };
    /**
     * @param {?} event
     * @return {?}
     */
    MetadatalistdetailsComponent.prototype.onSelectedPkgChanged = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        console.log(event);
        this.mdItem.selectedPkg = event.selectedRows[0].dataItem;
        this.selectedPkgChanged.emit(this.mdItem);
    };
    MetadatalistdetailsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-metadatalistdetails',
                    template: "<div class=\"gridDetails\">\r\n  <kendo-grid\r\n    [data]=\"detailView\"\r\n    [pageSize]=\"5\"\r\n    [skip]=\"detailSkip\"\r\n    [pageable]=\"true\"\r\n    [scrollable]=\"'none'\"\r\n    [selectable]=\"this.selectableSetting\"\r\n    (selectionChange)=onSelectedPkgChanged($event)\r\n    (pageChange)=\"detailPageChange($event)\"\r\n    [navigable]=\"true\"\r\n    kendoGridFocusable>\r\n    <kendo-grid-column field=\"name\" *ngIf=\"this.isMavenEnable === false || this.codelanguage === dotnet\" title=\"Nuget\u540D\u79F0\" width=\"120\">\r\n    </kendo-grid-column>\r\n    <kendo-grid-column field=\"version\" *ngIf=\"this.isMavenEnable === false || this.codelanguage === dotnet\" title=\"Nuget\u7248\u672C\" width=\"120\">\r\n    </kendo-grid-column>\r\n    <kendo-grid-column field=\"groupId\" *ngIf=\"this.isMavenEnable === true && this.codelanguage === java\" title=\"GroupId\" width=\"120\">\r\n    </kendo-grid-column>\r\n    <kendo-grid-column field=\"artifactId\" *ngIf=\"this.isMavenEnable === true && this.codelanguage === java\" title=\"ArtifactId\" width=\"120\">\r\n    </kendo-grid-column>\r\n    <kendo-grid-column field=\"version\" *ngIf=\"this.isMavenEnable === true && this.codelanguage === java\" title=\"Maven\u7248\u672C\" width=\"120\">\r\n    </kendo-grid-column>\r\n  </kendo-grid>\r\n</div>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    MetadatalistdetailsComponent.ctorParameters = function () { return []; };
    MetadatalistdetailsComponent.propDecorators = {
        isMavenEnable: [{ type: Input }],
        mdItem: [{ type: Input }],
        selectedPkgChanged: [{ type: Output }],
        codelanguage: [{ type: Input }]
    };
    return MetadatalistdetailsComponent;
}());
export { MetadatalistdetailsComponent };
if (false) {
    /** @type {?} */
    MetadatalistdetailsComponent.prototype.selectableSetting;
    /** @type {?} */
    MetadatalistdetailsComponent.prototype.dotnet;
    /** @type {?} */
    MetadatalistdetailsComponent.prototype.java;
    /** @type {?} */
    MetadatalistdetailsComponent.prototype.detailSkip;
    /** @type {?} */
    MetadatalistdetailsComponent.prototype.detailView;
    /** @type {?} */
    MetadatalistdetailsComponent.prototype.pageSize;
    /** @type {?} */
    MetadatalistdetailsComponent.prototype.isMavenEnable;
    /** @type {?} */
    MetadatalistdetailsComponent.prototype.mdItem;
    /** @type {?} */
    MetadatalistdetailsComponent.prototype.selectedPkgChanged;
    /** @type {?} */
    MetadatalistdetailsComponent.prototype.dataItem;
    /** @type {?} */
    MetadatalistdetailsComponent.prototype.codelanguage;
}
