// @farris/ui-datepicker/esm2015/lib/components/multi-day-view/multi-days.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
export class MultiSelectedDaysComponent {
    constructor() {
        this.locales = {};
        this.selectedDates = [];
        this.removed = new EventEmitter();
        this.clear = new EventEmitter();
        this.clicked = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
    }
    /**
     * @param {?} $event
     * @param {?} item
     * @return {?}
     */
    onRemove($event, item) {
        $event.stopPropagation();
        this.removed.emit(item);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClear($event) {
        $event.stopPropagation();
        this.clear.emit();
    }
    /**
     * @param {?} $event
     * @param {?} item
     * @return {?}
     */
    onSelectDateClick($event, item) {
        $event.stopPropagation();
        this.clicked.emit(item);
    }
}
MultiSelectedDaysComponent.decorators = [
    { type: Component, args: [{
                selector: 'multi-days',
                template: `
        <div class="d-flex mb-2 f-datepicker-multi-selectdays-header">
            <span style="font-size: 14px;" class="ml-3">{{ locales?.selected.split(',')[0] || '已选'}} <span style="font-weight: bold;font-size: 17px;">{{selectedDates.length}}</span> {{locales?.selected.split(',')[1] ||'天'}}</span>
            <span class="f-datepicker-multi-selectdays-clear" (click)="onClear($event)">{{ locales?.clearSelections || '清空'}}</span>
        </div>
        <div #selecteditems style="height: 230px;overflow-y: auto;" class="ml-3 multi-date-select win11Scroll nobtn" (wheel)="$event.stopPropagation()">
            <ul [style.width.px]="167">
                <li class="d-flex align-items-center f-datepicker-multi-selectdays" *ngFor="let d of selectedDates" (click)="onSelectDateClick($event, d)">
                    {{d.value}} <span class="f-icon f-icon-close remove" title="{{ locales?.delete || '删除'}}" (click)="onRemove($event, d)"></span>
                </li>
            </ul>
        </div>
    `,
                styles: [`
        .f-datepicker-multi-selectdays-header{
            height: 41px;border-bottom: 1px solid #E4E7EF;line-height: 40px;justify-content: space-between;
        }

        .f-datepicker-multi-selectdays {
            justify-content: space-between;color: #2D2F33;height: 30px;padding: 0 5px;border-radius: 6px;
            cursor: pointer;
        }

        .f-datepicker-multi-selectdays:hover {
            background: #F7F8FB;
        }

        .f-datepicker-multi-selectdays .remove {
            font-size: 12px;color: #C7CFDD;cursor: pointer;
        }

        .f-datepicker-multi-selectdays .remove:hover {
            color: #5f96f5;
        }

        .f-datepicker-multi-selectdays-clear {
            font-size: 13px;margin-right: 10px; color: #CCD1DD; cursor: pointer;
        }
        .f-datepicker-multi-selectdays-clear:hover {
            color: #5f96f5;
        }

        .mr14 {
            margin-right: 14px;
        }
        `]
            }] }
];
/** @nocollapse */
MultiSelectedDaysComponent.ctorParameters = () => [];
MultiSelectedDaysComponent.propDecorators = {
    locales: [{ type: Input }],
    selectedDates: [{ type: Input }],
    removed: [{ type: Output }],
    clear: [{ type: Output }],
    clicked: [{ type: Output }],
    selecteditemsEl: [{ type: ViewChild, args: ['selecteditems',] }]
};
if (false) {
    /** @type {?} */
    MultiSelectedDaysComponent.prototype.locales;
    /** @type {?} */
    MultiSelectedDaysComponent.prototype.selectedDates;
    /** @type {?} */
    MultiSelectedDaysComponent.prototype.removed;
    /** @type {?} */
    MultiSelectedDaysComponent.prototype.clear;
    /** @type {?} */
    MultiSelectedDaysComponent.prototype.clicked;
    /** @type {?} */
    MultiSelectedDaysComponent.prototype.selecteditemsEl;
}
