// @farris/ui-shortcuts/esm5/lib/types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function ShortcutBase() { }
if (false) {
    /**
     * Description for the command can be used for rendering help menu.
     * @type {?|undefined}
     */
    ShortcutBase.prototype.description;
    /**
     * How much time to throttle in ms.
     * @type {?|undefined}
     */
    ShortcutBase.prototype.throttleTime;
    /**
     * textarea, select and input are ignored by default, this is used to override
     * this behavior.
     * allow in node names, accepts: ["TEXTAREA", "SELECT", "INPUT"]
     * @type {?|undefined}
     */
    ShortcutBase.prototype.allowIn;
    /**
     * Label, can be used for grouping commands.
     * @type {?|undefined}
     */
    ShortcutBase.prototype.label;
    /**
     * 只有当目标获得焦点后，快捷键才会触发命令.
     * @type {?|undefined}
     */
    ShortcutBase.prototype.target;
    /**
     * Prevent browser default, default: false
     * @type {?|undefined}
     */
    ShortcutBase.prototype.preventDefault;
    /**
     * callback to be called when shortcut is pressed.
     * @param {?} event - the event out
     * @return {?}
     */
    ShortcutBase.prototype.command = function (event) { };
}
/**
 * @record
 */
export function Shortcut() { }
if (false) {
    /** @type {?} */
    Shortcut.prototype.key;
    /** @type {?|undefined} */
    Shortcut.prototype.component;
}
/**
 * @record
 */
export function ShortcutInput() { }
if (false) {
    /** @type {?} */
    ShortcutInput.prototype.key;
}
/**
 * @record
 */
export function ParsedShortcut() { }
if (false) {
    /** @type {?} */
    ParsedShortcut.prototype.predicates;
    /** @type {?|undefined} */
    ParsedShortcut.prototype.priority;
    /** @type {?|undefined} */
    ParsedShortcut.prototype.event;
}
/**
 * @record
 */
export function ShortcutEventOutput() { }
if (false) {
    /**
     * The browser keyboard event
     * @type {?}
     */
    ShortcutEventOutput.prototype.event;
    /** @type {?} */
    ShortcutEventOutput.prototype.key;
}
