// @farris/ui-shortcuts/esm5/lib/utils.js
var _this = this;
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} x
 * @return {?}
 */
export function isFunction(x) {
    return typeof x === 'function';
}
/** @type {?} */
export var any = (/**
 * @param {?} fn
 * @param {?} list
 * @return {?}
 */
function (fn, list) {
    /** @type {?} */
    var idx = 0;
    while (idx < list.length) {
        if (fn(list[idx])) {
            return true;
        }
        idx += 1;
    }
    return false;
});
/** @type {?} */
export var identity = (/**
 * @param {?} x
 * @return {?}
 */
function (x) { return x; });
/** @type {?} */
export var isNill = (/**
 * @param {?} x
 * @return {?}
 */
function (x) { return x == null; });
/** @type {?} */
export var difference = (/**
 * @param {?} first
 * @param {?} second
 * @return {?}
 */
function (first, second) {
    return first.filter((/**
     * @param {?} item
     * @return {?}
     */
    function (item) { return !((/** @type {?} */ (second))).some((/**
     * @param {?} secondItem
     * @return {?}
     */
    function (secondItem) { return secondItem == item; })); }));
});
/** @type {?} */
export var allPass = (/**
 * @param {?} preds
 * @return {?}
 */
function (preds) { return (/**
 * @param {...?} args
 * @return {?}
 */
function () {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    /** @type {?} */
    var idx = 0;
    /** @type {?} */
    var len = preds.length;
    while (idx < len) {
        if (!preds[idx].apply(_this, args)) {
            return false;
        }
        idx += 1;
    }
    return true;
}); });
