// @gsp-sys/rtf-ui/esm5/lib/rtf-app-help/app-module-tree/app-module-tree.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { from } from 'rxjs';
var AppModuleTreeComponent = /** @class */ (function () {
    function AppModuleTreeComponent() {
        /**
         * 菜单明细类
         */
        this.appTreeData = [];
        this.selectKeys = [];
        /**
         * 功能菜单模块变化后后
         */
        this.appTreeChanged = new EventEmitter();
        /**
         * 树节点展开后的输出
         */
        this.appTreeExpanded = new EventEmitter();
        /**
         * 树控件展开的节点值
         */
        this.expandedKeys = [];
    }
    /**
     * 变化情况
     * @param changes 变化
     */
    /**
     * 变化情况
     * @param {?} changes 变化
     * @return {?}
     */
    AppModuleTreeComponent.prototype.ngOnChanges = /**
     * 变化情况
     * @param {?} changes 变化
     * @return {?}
     */
    function (changes) {
        var _this = this;
        // 初始化变化取消
        // 未包含appTreeData的变化
        if (!changes['appTreeData']) {
            return;
        }
        // 无效的值不予处理
        if (!changes['appTreeData'].currentValue ||
            changes['appTreeData'].currentValue.length === 0) {
            return;
        }
        // 获取到菜单数据
        from(this.appTreeData).subscribe((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (!item.ParentId || item.ParentId === '0') {
                item.ParentId = undefined;
                if (item.Layer === 1) {
                    _this.expandedKeys.push(item.Id);
                }
            }
        }));
        /*
         // 默认业务对象的处理
         if (!isNullOrUndefined(this.bizObjectId) && this.bizObjectId !== '') {
            // 树焦点行
            let curBizObjectId = this.bizObjectId;
            while (!isNullOrUndefined(curBizObjectId)) {
                this.expandedKeys.push(curBizObjectId);
                // 找到当前节点的付家店
                curBizObjectId = this.data.find(item => item['id'] === curBizObjectId)['parentID'];
            }
            // 右侧操作变化
            this.selectKeys = [this.bizObjectId];
            setTimeout(() => {
                this.bizObjectChanged.emit(this.bizObjectId);
            }, 80);
        }
        */
    };
    /**
     * @return {?}
     */
    AppModuleTreeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * 选中行改变后触发事件
     */
    /**
     * 选中行改变后触发事件
     * @param {?} args
     * @return {?}
     */
    AppModuleTreeComponent.prototype.selectChange = /**
     * 选中行改变后触发事件
     * @param {?} args
     * @return {?}
     */
    function (args) {
        /** @type {?} */
        var curApp = args.dataItem;
        // 首次触发
        if (this.selectKeys == null || this.selectKeys.length === 0) {
            this.appTreeChanged.emit(curApp);
        }
        else {
            // 这个事件里还不会对selectedkeys数组赋值，是上一个状态的
            /** @type {?} */
            var lastAppId = this.selectKeys[0];
            /** @type {?} */
            var curAppId = curApp.Id;
            // 变化的情况下触发
            if (lastAppId !== curAppId) {
                this.appTreeChanged.emit(curApp);
            }
        }
    };
    /**
     * @param {?} arg
     * @return {?}
     */
    AppModuleTreeComponent.prototype.onExpand = /**
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        console.log('菜单帮助左侧树展开');
        if (!arg || !arg.dataItem) {
            return;
        }
        this.appTreeExpanded.emit(arg.dataItem);
    };
    AppModuleTreeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-module-tree',
                    template: "<div class=\"d-flex flex-column flex-fill pt-2 pb-2\" style=\"height: 100%\">\r\n    <kendo-treeview [nodes]=\"appTreeData\" textField=\"Name\" kendoTreeViewExpandable kendoTreeViewFlatDataBinding\r\n        idField=\"Id\" parentIdField=\"ParentId\" kendoTreeViewSelectable [selectBy]=\"'Id'\"\r\n        [(selectedKeys)]=\"this.selectKeys\" [(expandedKeys)]=\"expandedKeys\" [expandBy]=\"'Id'\"\r\n        (expand)=\"this.onExpand($event)\" (selectionChange)=\"selectChange($event)\">\r\n    </kendo-treeview>\r\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    AppModuleTreeComponent.ctorParameters = function () { return []; };
    AppModuleTreeComponent.propDecorators = {
        appTreeData: [{ type: Input }],
        bizObjectId: [{ type: Input }],
        appTreeChanged: [{ type: Output }],
        appTreeExpanded: [{ type: Output }]
    };
    return AppModuleTreeComponent;
}());
export { AppModuleTreeComponent };
if (false) {
    /**
     * 菜单明细类
     * @type {?}
     */
    AppModuleTreeComponent.prototype.appTreeData;
    /**
     * 业务对象
     *
     * @type {?}
     */
    AppModuleTreeComponent.prototype.bizObjectId;
    /** @type {?} */
    AppModuleTreeComponent.prototype.selectKeys;
    /** @type {?} */
    AppModuleTreeComponent.prototype.currentid;
    /**
     * 功能菜单模块变化后后
     * @type {?}
     */
    AppModuleTreeComponent.prototype.appTreeChanged;
    /**
     * 树节点展开后的输出
     * @type {?}
     */
    AppModuleTreeComponent.prototype.appTreeExpanded;
    /**
     * 树控件展开的节点值
     * @type {?}
     */
    AppModuleTreeComponent.prototype.expandedKeys;
}
