// @gsp-wf/wf-process-editor/esm5/lib/views/multi-instance-loop-characteristics/multi-instance-loop-characteristics.entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MultiInstanceLoopCharacteristics = /** @class */ (function () {
    function MultiInstanceLoopCharacteristics() {
        this.ClrTypeID = MultiInstanceLoopCharacteristics.Clr_Type_ID;
        this.enable = false;
        this.sequential = false;
        this.loopCardinality = null;
        this.completionCondition = null;
        this.completionConditionRules = new Array();
    }
    MultiInstanceLoopCharacteristics.Clr_Type_ID = 'wf.MultiInstanceLoopCharacteristics';
    return MultiInstanceLoopCharacteristics;
}());
export { MultiInstanceLoopCharacteristics };
if (false) {
    /** @type {?} */
    MultiInstanceLoopCharacteristics.Clr_Type_ID;
    /** @type {?} */
    MultiInstanceLoopCharacteristics.prototype.ClrTypeID;
    /** @type {?} */
    MultiInstanceLoopCharacteristics.prototype.enable;
    /** @type {?} */
    MultiInstanceLoopCharacteristics.prototype.sequential;
    /** @type {?} */
    MultiInstanceLoopCharacteristics.prototype.loopCardinality;
    /** @type {?} */
    MultiInstanceLoopCharacteristics.prototype.completionCondition;
    /** @type {?} */
    MultiInstanceLoopCharacteristics.prototype.completionConditionRules;
}
