// @farris/ui-feature-editor/esm5/lib/lookup.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { Injectable } from '@angular/core';
var LookupService = /** @class */ (function () {
    function LookupService(http) {
        this.http = http;
    }
    /**
     * @param {?} url
     * @param {?} params
     * @return {?}
     */
    LookupService.prototype.getData = /**
     * @param {?} url
     * @param {?} params
     * @return {?}
     */
    function (url, params) {
        /** @type {?} */
        var urls = url.split(':');
        /** @type {?} */
        var realUrl = '/api/scm/scmfnd/v1.0/scmcommoncmpservice/service/gethelpdata';
        /** @type {?} */
        var helpID = urls[0] === '1' ? '9e2dca82-0aa2-4394-b26d-cb91164f44d5' : '344ffc53-7167-48d1-866b-a84083404e9e';
        /** @type {?} */
        var filter = 'setid:' + urls[1] + '&' + 'state_isenabled:1';
        /** @type {?} */
        var httpParams = {};
        if (params) {
            if (params.pageIndex) {
                httpParams['pageIndex'] = JSON.stringify(params.pageIndex - 0);
            }
            if (params.pageSize) {
                httpParams['pageSize'] = JSON.stringify(params.pageSize - 0);
            }
            if (params.condition) {
                httpParams['condition'] = JSON.stringify(params.condition);
            }
            if (params.searchValue) {
                httpParams['search'] = params.searchValue;
            }
        }
        return this.http.put(realUrl, {
            helpID: helpID,
            queryParam: JSON.stringify(httpParams),
            filterStr: filter
        }).pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return JSON.parse(data.returnValue); })));
    };
    LookupService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    LookupService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return LookupService;
}());
export { LookupService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    LookupService.prototype.http;
}
