// @gsp-wf/rtdevkit/esm2015/lib/entity/complete-response.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 指定参与者信息：需要指定参与者的活动、待指定用户列表
 */
export class AssigneeInfo {
}
if (false) {
    /**
     * 工作项ID
     * @type {?}
     */
    AssigneeInfo.prototype.workItemId;
    /**
     * 参与者列表
     * @type {?}
     */
    AssigneeInfo.prototype.assigneeIds;
    /**
     * 参与者用户列表
     * @type {?}
     */
    AssigneeInfo.prototype.assigneeUsers;
    /**
     * 是否允许指派
     * @type {?}
     */
    AssigneeInfo.prototype.selectAssignee;
    /**
     * 是否多选
     * @type {?}
     */
    AssigneeInfo.prototype.multiSelect;
    /**
     * 是否允许全员指派
     * @type {?}
     */
    AssigneeInfo.prototype.allowGlobalAssign;
    /**
     * 指派范围
     * @type {?}
     */
    AssigneeInfo.prototype.organizationScope;
}
/**
 * 加签移交征询范围设置
 */
export class OrganizationScope {
}
if (false) {
    /** @type {?} */
    OrganizationScope.prototype.id;
    /** @type {?} */
    OrganizationScope.prototype.code;
    /** @type {?} */
    OrganizationScope.prototype.name;
    /** @type {?} */
    OrganizationScope.prototype.parentId;
    /** @type {?} */
    OrganizationScope.prototype.path;
    /** @type {?} */
    OrganizationScope.prototype.detail;
    /** @type {?} */
    OrganizationScope.prototype.layer;
    /** @type {?} */
    OrganizationScope.prototype.includeChildren;
    /** @type {?} */
    OrganizationScope.prototype.parentsUntil;
    /** @type {?} */
    OrganizationScope.prototype.principalType;
}
/**
 * 根据流程指定下一步参与者参数
 */
export class AssigneeInfoByProcInstIdParameter {
}
if (false) {
    /**
     * 流程实例ID
     * @type {?}
     */
    AssigneeInfoByProcInstIdParameter.prototype.procInstId;
    /**
     * 指定参与者信息
     * @type {?}
     */
    AssigneeInfoByProcInstIdParameter.prototype.assigneeInfos;
}
/**
 * 根据流程指定下一步参与者参数
 */
export class AssigneeInfoByWorkItemIdParameter {
}
if (false) {
    /** @type {?} */
    AssigneeInfoByWorkItemIdParameter.prototype.workItemId;
    /** @type {?} */
    AssigneeInfoByWorkItemIdParameter.prototype.assigneeIds;
    /** @type {?} */
    AssigneeInfoByWorkItemIdParameter.prototype.comment;
}
/**
 * 完成工作项返回结果
 */
export class CompleteResponse {
}
if (false) {
    /** @type {?} */
    CompleteResponse.prototype.dataId;
    /** @type {?} */
    CompleteResponse.prototype.needInteraction;
    /** @type {?} */
    CompleteResponse.prototype.nextAssigneeInfos;
    /** @type {?} */
    CompleteResponse.prototype.procInstId;
    /** @type {?} */
    CompleteResponse.prototype.nextFlowNodes;
    /** @type {?} */
    CompleteResponse.prototype.procDefs;
}
export class User {
}
if (false) {
    /** @type {?} */
    User.prototype.id;
    /** @type {?} */
    User.prototype.code;
    /** @type {?} */
    User.prototype.name;
    /** @type {?} */
    User.prototype.note;
    /** @type {?} */
    User.prototype.orgId;
    /** @type {?} */
    User.prototype.orgCode;
    /** @type {?} */
    User.prototype.orgName;
    /** @type {?} */
    User.prototype.orgNamePath;
    /** @type {?} */
    User.prototype.orderNum;
}
export class ProcessDefinition {
}
if (false) {
    /** @type {?} */
    ProcessDefinition.prototype.procDefId;
    /** @type {?} */
    ProcessDefinition.prototype.procDefKey;
    /** @type {?} */
    ProcessDefinition.prototype.procDefName;
    /** @type {?} */
    ProcessDefinition.prototype.suspensionState;
    /** @type {?} */
    ProcessDefinition.prototype.version;
    /** @type {?} */
    ProcessDefinition.prototype.description;
}
