// @farris/ide-devkit/esm2015/lib/components/navbar/navbar.state.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
export class NavbarState {
    constructor() {
        this.dropdownInstances = [];
        this.menuItems = [];
        this.subscriptions = [];
        // 当前下拉菜单的子菜单
        this.onItemClick = new EventEmitter();
    }
    /**
     * @param {?} dropdown
     * @return {?}
     */
    addDropdown(dropdown) {
        if (!this.dropdownInstances.find((item) => item === dropdown)) {
            this.dropdownInstances.push(dropdown);
            this.subscribeDropdown(dropdown);
            dropdown.onShown.subscribe(() => setTimeout(() => {
                dropdown.childDropdowns.forEach((item) => this.addDropdown(item));
            }));
        }
    }
    /**
     * @param {?} menuItem
     * @return {?}
     */
    addMenuItem(menuItem) {
        if (!this.menuItems.find((item) => item === menuItem)) {
            this.menuItems.push(menuItem);
            this.subscribeMenuItem(menuItem);
        }
    }
    /**
     * @private
     * @param {?} dropdown
     * @return {?}
     */
    subscribeDropdown(dropdown) {
        // 在下拉菜单显示、隐藏时，记录下显示状态。
        // this.subscriptions.push(
        //   dropdown.onShown.subscribe((value) => {
        //     debugger;
        //     if(dropdown.isSubMenu) {
        //
        //     }
        //   })
        // );
        //
        // this.subscriptions.push(
        //   dropdown.onHidden.subscribe((value) => {
        //
        //   })
        // );
        this.subscriptions.push(dropdown.state.onClick.subscribe((event) => {
            if (dropdown.level === 1) {
                if (dropdown.isOpen && this.level1Dropdown === dropdown) {
                    this.innerHideL1Dropdown();
                }
                else {
                    this.innerShowL1Dropdown(dropdown);
                }
            }
            else if (dropdown.level === 2) {
                if (!dropdown.isOpen || this.level2Dropdown !== dropdown) {
                    this.innerShowL2Dropdown(dropdown);
                }
            }
            event.stopPropagation();
        }));
        this.subscriptions.push(dropdown.state.onHover.subscribe(() => {
            if (!this.level1Dropdown) {
                // 菜单没展开，鼠标悬停没有效果
                return;
            }
            this.show(dropdown);
        }));
    }
    /**
     * @private
     * @param {?} menuItem
     * @return {?}
     */
    subscribeMenuItem(menuItem) {
        this.subscriptions.push(menuItem.onHover.subscribe(() => {
            if (!this.level1Dropdown) {
                return;
            }
            if (this.level1Dropdown.containsItem(menuItem)) {
                this.innerHideL2Dropdown();
            }
        }));
        this.subscriptions.push(menuItem.onClick.subscribe((command) => {
            this.onItemClick.emit(command);
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.subscriptions.forEach((sub) => {
            sub.unsubscribe();
        });
    }
    /**
     * @return {?}
     */
    hideCurrentDropdown() {
        if (this.level2Dropdown) {
            this.innerHideL2Dropdown();
        }
        else {
            this.innerHideL1Dropdown();
        }
    }
    /**
     * @return {?}
     */
    hideAllDropdown() {
        this.innerHideL2Dropdown();
        this.innerHideL1Dropdown();
    }
    /**
     * @param {?} dropdown
     * @return {?}
     */
    show(dropdown) {
        if (dropdown.level === 1) {
            if (dropdown !== this.level1Dropdown) {
                this.innerShowL1Dropdown(dropdown);
            }
            this.innerHideL2Dropdown();
        }
        else if (dropdown.level === 2) {
            if (dropdown !== this.level2Dropdown) {
                this.innerShowL2Dropdown(dropdown);
            }
        }
        ///////////////////////////////
        /*if (this.isMenuShown && this.isSubMenuShown && this.parentDropdown && this.parentDropdown.containsItem(dropdown)) {
          // 要显示的是子菜单，不是当前子菜单
          if (this.currentDropdown && this.currentDropdown !== dropdown) {
            this.currentDropdown.hide();
          }
          dropdown.show();
          this.currentDropdown = dropdown;
          this.isSubMenuShown = true;
        } else if (this.isMenuShown && !this.isSubMenuShown && this.currentDropdown && this.currentDropdown.containsItem(dropdown)) {
    
        } else {
          // 主菜单
          if (this.isSubMenuShown && this.currentDropdown && this.parentDropdown) {
            this.currentDropdown.hide();
            this.currentDropdown = null;
            this.isSubMenuShown = false;
          }
          if (this.isMenuShown && this.currentDropdown && this.currentDropdown !== dropdown) {
            this.currentDropdown.hide();
          }
          dropdown.show();
          this.currentDropdown = dropdown;
          this.isMenuShown = true;
        }*/
    }
    /**
     * @param {?} dropdown
     * @return {?}
     */
    hide(dropdown) {
        // if
    }
    /**
     * @private
     * @param {?} target
     * @return {?}
     */
    innerShowL1Dropdown(target) {
        if (this.level1Dropdown) {
            this.level1Dropdown.hide();
        }
        target.show();
        this.level1Dropdown = target;
    }
    /**
     * @private
     * @return {?}
     */
    innerHideL1Dropdown() {
        if (this.level1Dropdown) {
            this.level1Dropdown.hide();
            this.level1Dropdown = null;
        }
    }
    /**
     * @private
     * @param {?} target
     * @return {?}
     */
    innerShowL2Dropdown(target) {
        if (this.level2Dropdown) {
            this.level2Dropdown.hide();
        }
        target.show();
        this.level2Dropdown = target;
    }
    /**
     * @private
     * @return {?}
     */
    innerHideL2Dropdown() {
        if (this.level2Dropdown) {
            this.level2Dropdown.hide();
            this.level2Dropdown = null;
        }
    }
}
NavbarState.decorators = [
    { type: Injectable }
];
/** @nocollapse */
NavbarState.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    NavbarState.prototype.dropdownInstances;
    /**
     * @type {?}
     * @private
     */
    NavbarState.prototype.menuItems;
    /**
     * @type {?}
     * @private
     */
    NavbarState.prototype.subscriptions;
    /**
     * @type {?}
     * @private
     */
    NavbarState.prototype.level1Dropdown;
    /**
     * @type {?}
     * @private
     */
    NavbarState.prototype.level2Dropdown;
    /** @type {?} */
    NavbarState.prototype.onItemClick;
    /** @type {?} */
    NavbarState.prototype.autoClose;
}
