// @farris/ide-devkit/esm2015/lib/components/navbar/navbar.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ContentChildren, Directive, HostListener, QueryList, ViewChildren } from '@angular/core';
import { NavbarState } from './navbar.state';
import { ElementRef } from '@angular/core';
import { BsDropdownConfig } from '../dropdown/dropdown.config';
import { BsDropdownDirective } from '../dropdown/index';
export class NavbarDirective {
    /**
     * @param {?} _elementRef
     * @param {?} _config
     * @param {?} _state
     */
    constructor(_elementRef, _config, _state) {
        this._elementRef = _elementRef;
        this._config = _config;
        this._state = _state;
        this._state.autoClose = this._config.autoClose;
        this.onItemClick = this._state.onItemClick;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onDocumentClick(event) {
        this._state.hideAllDropdown();
    }
    /**
     * @return {?}
     */
    onEsc() {
        if (this._state.autoClose) {
            this._state.hideCurrentDropdown();
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.childDropdowns.forEach((item) => this._state.addDropdown(item));
    }
}
NavbarDirective.decorators = [
    { type: Directive, args: [{
                selector: '[navbar]',
                exportAs: 'navbar',
                providers: [NavbarState]
            },] }
];
/** @nocollapse */
NavbarDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: BsDropdownConfig },
    { type: NavbarState }
];
NavbarDirective.propDecorators = {
    childDropdowns: [{ type: ContentChildren, args: [BsDropdownDirective,] }],
    childDropdowns2: [{ type: ViewChildren, args: [BsDropdownDirective,] }],
    onDocumentClick: [{ type: HostListener, args: ['document:click', ['$event'],] }],
    onEsc: [{ type: HostListener, args: ['keyup.esc',] }]
};
if (false) {
    /** @type {?} */
    NavbarDirective.prototype.onItemClick;
    /** @type {?} */
    NavbarDirective.prototype.childDropdowns;
    /** @type {?} */
    NavbarDirective.prototype.childDropdowns2;
    /**
     * @type {?}
     * @private
     */
    NavbarDirective.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    NavbarDirective.prototype._config;
    /**
     * @type {?}
     * @private
     */
    NavbarDirective.prototype._state;
}
