// @edp-bif/runtime-api/esm5/lib/bill-tracker/args/CanTrackingResult.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/args/CanTrackingResult.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '@edp-aif/common-api';
import { CanTrackingInfo } from '../../entity/tracking/CanTrackingInfo';
var CanTrackingResult = /** @class */ (function (_super) {
    tslib_1.__extends(CanTrackingResult, _super);
    function CanTrackingResult() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    CanTrackingResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.billTrackingSettings = null;
        this.voName = null;
    };
    /**
     * @return {?}
     */
    CanTrackingResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billTrackingSettings", this.billTrackingSettings);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voName", this.voName);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CanTrackingResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.billTrackingSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billTrackingSettings");
        this.voName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voName");
    };
    return CanTrackingResult;
}(CanTrackingInfo));
export { CanTrackingResult };
if (false) {
    /**
     * 现为表组件展示字段，原为所有配置信息
     * @type {?}
     */
    CanTrackingResult.prototype.billTrackingSettings;
    /** @type {?} */
    CanTrackingResult.prototype.voName;
}
