// @edp-bif/runtime-api/esm5/lib/bill-tracker/args/EntityDataResult.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/args/EntityDataResult.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper, BillSchema, BillData, } from '@edp-aif/common-api';
var EntityDataResult = /** @class */ (function (_super) {
    tslib_1.__extends(EntityDataResult, _super);
    function EntityDataResult() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    EntityDataResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.billSchema = null;
        this.billData = null;
    };
    /**
     * @return {?}
     */
    EntityDataResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billSchema", this.billSchema);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billData", this.billData);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    EntityDataResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.billSchema = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "billSchema", BillSchema);
        this.billData = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "billData", BillData);
    };
    return EntityDataResult;
}(JsonConvertableImpl));
export { EntityDataResult };
if (false) {
    /** @type {?} */
    EntityDataResult.prototype.billSchema;
    /** @type {?} */
    EntityDataResult.prototype.billData;
}
