// @edp-bif/runtime-api/esm5/lib/bill-tracker/args/GetChildrenEntityDataArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/args/GetChildrenEntityDataArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from '@edp-aif/common-api';
var GetChildrenEntityDataArgs = /** @class */ (function (_super) {
    tslib_1.__extends(GetChildrenEntityDataArgs, _super);
    function GetChildrenEntityDataArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    GetChildrenEntityDataArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.voId = null;
        this.voMetadataId = null;
        this.suCode = null;
        this.nodeCodes = null;
        this.hierachyIdList = null;
    };
    /**
     * @return {?}
     */
    GetChildrenEntityDataArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voMetadataId", this.voMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "nodeCodes", BpmnModelJsonHelper.StringListToJson(this.nodeCodes));
        BpmnModelJsonHelper.SetAttributeValue(obj, "hierachyIdList", BpmnModelJsonHelper.StringListToJson(this.hierachyIdList));
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    GetChildrenEntityDataArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
        this.voMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voMetadataId");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.nodeCodes = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "nodeCodes"));
        this.hierachyIdList = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "hierachyIdList"));
    };
    return GetChildrenEntityDataArgs;
}(JsonConvertableImpl));
export { GetChildrenEntityDataArgs };
if (false) {
    /** @type {?} */
    GetChildrenEntityDataArgs.prototype.voId;
    /** @type {?} */
    GetChildrenEntityDataArgs.prototype.voMetadataId;
    /** @type {?} */
    GetChildrenEntityDataArgs.prototype.suCode;
    /**
     * @type {?}
     * @private
     */
    GetChildrenEntityDataArgs.prototype.nodeCodes;
    /**
     * @type {?}
     * @private
     */
    GetChildrenEntityDataArgs.prototype.hierachyIdList;
}
