// @edp-bif/runtime-api/esm5/lib/bill-tracker/args/GetJointQueryParametersArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/args/GetJointQueryParametersArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { JointQueryParametersContent } from "../../entity/tracking/JointQueryParametersContent";
var GetJointQueryParametersArgs = /** @class */ (function (_super) {
    tslib_1.__extends(GetJointQueryParametersArgs, _super);
    function GetJointQueryParametersArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.contents = [];
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    GetJointQueryParametersArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.trackingSourceType = null;
        this.suCode = null;
        this.schemaId = null;
        this.dataId = null;
        this.contents = [];
        this.curYear = null;
    };
    /**
     * @return {?}
     */
    GetJointQueryParametersArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "trackingSourceType", this.trackingSourceType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaId", this.schemaId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dataId", this.dataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "contents", this.contents);
        BpmnModelJsonHelper.SetAttributeValue(obj, "curYear", this.curYear);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    GetJointQueryParametersArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.trackingSourceType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "trackingSourceType");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.schemaId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaId");
        this.dataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dataId");
        this.contents = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "contents", JointQueryParametersContent);
        this.curYear = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "curYear");
    };
    return GetJointQueryParametersArgs;
}(JsonConvertableImpl));
export { GetJointQueryParametersArgs };
if (false) {
    /** @type {?} */
    GetJointQueryParametersArgs.prototype.trackingSourceType;
    /** @type {?} */
    GetJointQueryParametersArgs.prototype.suCode;
    /** @type {?} */
    GetJointQueryParametersArgs.prototype.schemaId;
    /** @type {?} */
    GetJointQueryParametersArgs.prototype.dataId;
    /** @type {?} */
    GetJointQueryParametersArgs.prototype.curYear;
    /** @type {?} */
    GetJointQueryParametersArgs.prototype.contents;
}
