// @gsp-wf/wf-process-model/esm2015/lib/model/participants/old/ActivityExecutorParticipant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { Participant } from "./Participant";
export class ActivityExecutorParticipant extends Participant {
    /**
     * @param {?} model
     * @param {?=} ActivityDefId
     * @param {?=} ActivityDefName
     */
    constructor(model, ActivityDefId, ActivityDefName) {
        super(model);
        this.ClrTypeID = ActivityExecutorParticipant.Clr_Type_ID;
        this.ActivityDefId = ActivityDefId;
        this.ActivityDefName = ActivityDefName;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "ActivityDefId", this.ActivityDefId);
        BpmnModelHelper.SetAttributeValue(obj, "ActivityDefName", this.ActivityDefName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.ActivityDefId = obj.ActivityDefId;
        this.ActivityDefName = obj.ActivityDefName;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ActivityExecutorParticipant(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const activityExecutorParticipant = new ActivityExecutorParticipant(this.Model);
        activityExecutorParticipant.Id = BpmnModelHelper.GenerateElementId();
        activityExecutorParticipant.ActivityDefId = this.ActivityDefId;
        activityExecutorParticipant.ActivityDefName = this.ActivityDefName;
        if (this.Condition) {
            activityExecutorParticipant.Condition = this.Condition.clone();
        }
        return activityExecutorParticipant;
    }
}
ActivityExecutorParticipant.Clr_Type_ID = 'wf.ActivityExecutorParticipant';
ActivityExecutorParticipant.CLR_TYPE_NAME = '活动办理人';
if (false) {
    /** @type {?} */
    ActivityExecutorParticipant.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    ActivityExecutorParticipant.CLR_TYPE_NAME;
    /** @type {?} */
    ActivityExecutorParticipant.prototype.ClrTypeID;
    /** @type {?} */
    ActivityExecutorParticipant.prototype.ActivityDefId;
    /** @type {?} */
    ActivityExecutorParticipant.prototype.ActivityDefName;
}
BpmnModelHelper.RegisterElementType(new ActivityExecutorParticipant(new BpmnModel()));
