// @gsp-wf/wf-process-model/esm2015/lib/model/participants/old/Participant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HumanPerformer } from "@edp-pmf/bpmn-model";
import { Condition } from "../../extension/condition-expression/Condition";
/**
 * @abstract
 */
export class Participant extends HumanPerformer {
    /**
     * @param {?} model
     * @param {?=} id
     * @param {?=} name
     */
    constructor(model, id, name) {
        super(model);
        this.ClrTypeID = "wf.Participant";
        this.MexTag = 'wf.Participant';
        this.Id = id;
        this.Name = name;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
    }
    /**
     * @protected
     * @return {?}
     */
    GetExtensionElements4ConvertTo() {
        /** @type {?} */
        let list = super.GetExtensionElements4ConvertTo();
        if (this.Condition) {
            this.Condition.MexTag = this.MexTag;
            list.push(this.Condition);
        }
        return list;
    }
    /**
     * @protected
     * @param {?} extensionElements
     * @return {?}
     */
    SetExtensionElements4LoadFrom(extensionElements) {
        super.SetExtensionElements4LoadFrom(extensionElements);
        for (const item of extensionElements) {
            if (item instanceof Condition) {
                /** @type {?} */
                const condition = (/** @type {?} */ (item));
                if (condition.MexTag === this.MexTag) {
                    this.Condition = condition;
                }
            }
        }
    }
}
if (false) {
    /** @type {?} */
    Participant.prototype.ClrTypeID;
    /** @type {?} */
    Participant.prototype.MexTag;
    /** @type {?} */
    Participant.prototype.Condition;
}
