// @gsp-wf/wf-process-model/esm2015/lib/model/participants/old/PositionParticipant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { Participant } from "./Participant";
export class PositionParticipant extends Participant {
    /**
     * @param {?} model
     * @param {?=} id
     * @param {?=} name
     */
    constructor(model, id, name) {
        super(model);
        this.ClrTypeID = PositionParticipant.Clr_Type_ID;
        /// <summary>
        /// 组织类型Code
        /// </summary>
        this.OrgType = 'Sys';
        this.PositionId = id;
        this.PositionName = name;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "PositionId", this.PositionId);
        BpmnModelHelper.SetAttributeValue(obj, "PositionName", this.PositionName);
        BpmnModelHelper.SetAttributeValue(obj, "PositionType", this.PositionType);
        BpmnModelHelper.SetAttributeValue(obj, "OrgId", this.OrgId);
        BpmnModelHelper.SetAttributeValue(obj, "OrgName", this.OrgName);
        BpmnModelHelper.SetAttributeValue(obj, "OrgType", this.OrgType);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.PositionId = obj.PositionId;
        this.PositionName = obj.PositionName;
        this.PositionType = obj.PositionType;
        this.OrgId = obj.OrgId;
        this.OrgName = obj.OrgName;
        this.OrgType = obj.OrgType;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new PositionParticipant(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const positionParticipant = new PositionParticipant(this.Model);
        positionParticipant.Id = BpmnModelHelper.GenerateElementId();
        positionParticipant.Name = this.Name;
        positionParticipant.PositionId = this.PositionId;
        positionParticipant.PositionName = this.PositionName;
        positionParticipant.PositionType = this.PositionType;
        positionParticipant.OrgType = this.OrgType;
        positionParticipant.OrgId = this.OrgId;
        positionParticipant.OrgName = this.OrgName;
        if (this.Condition) {
            positionParticipant.Condition = this.Condition.clone();
        }
        return positionParticipant;
    }
}
PositionParticipant.Clr_Type_ID = 'wf.PositionParticipant';
PositionParticipant.CLR_TYPE_NAME = '岗位';
if (false) {
    /** @type {?} */
    PositionParticipant.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    PositionParticipant.CLR_TYPE_NAME;
    /** @type {?} */
    PositionParticipant.prototype.ClrTypeID;
    /** @type {?} */
    PositionParticipant.prototype.PositionId;
    /** @type {?} */
    PositionParticipant.prototype.PositionName;
    /** @type {?} */
    PositionParticipant.prototype.PositionType;
    /** @type {?} */
    PositionParticipant.prototype.OrgId;
    /** @type {?} */
    PositionParticipant.prototype.OrgName;
    /** @type {?} */
    PositionParticipant.prototype.OrgType;
}
BpmnModelHelper.RegisterElementType(new PositionParticipant(new BpmnModel()));
