// @gsp-wf/wf-process-model/esm2015/lib/model/participants/old/ReportingLineParticipant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { Participant } from "./Participant";
export class ReportingLineParticipant extends Participant {
    /**
     * @param {?} model
     * @param {?=} RefUserId
     * @param {?=} RefEmployeeId
     * @param {?=} ReportingReferObject
     * @param {?=} RefOrgId
     * @param {?=} RefOrgName
     * @param {?=} ReportingLineDefId
     * @param {?=} ReportingLineDefName
     * @param {?=} Dimension
     */
    constructor(model, RefUserId, RefEmployeeId, ReportingReferObject, RefOrgId, RefOrgName, ReportingLineDefId, ReportingLineDefName, Dimension) {
        super(model);
        this.ClrTypeID = ReportingLineParticipant.Clr_Type_ID;
        this.RefUserId = RefUserId;
        this.RefEmployeeId = RefEmployeeId;
        this.ReportingReferObject = ReportingReferObject;
        this.RefOrgId = RefOrgId;
        this.RefOrgName = RefOrgName;
        this.ReportingLineDefId = ReportingLineDefId;
        this.ReportingLineDefName = ReportingLineDefName;
        this.Dimension = Dimension;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "RefUserId", this.RefUserId);
        BpmnModelHelper.SetAttributeValue(obj, "RefEmployeeId", this.RefEmployeeId);
        BpmnModelHelper.SetAttributeValue(obj, "ReportingReferObject", this.ReportingReferObject);
        BpmnModelHelper.SetAttributeValue(obj, "Dimension", this.Dimension);
        BpmnModelHelper.SetAttributeValue(obj, "RefOrgId", this.RefOrgId);
        BpmnModelHelper.SetAttributeValue(obj, "RefOrgName", this.RefOrgName);
        BpmnModelHelper.SetAttributeValue(obj, "ReportingLineDefId", this.ReportingLineDefId);
        BpmnModelHelper.SetAttributeValue(obj, "ReportingLineDefName", this.ReportingLineDefName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.RefUserId = obj.RefUserId;
        this.RefEmployeeId = obj.RefEmployeeId;
        this.ReportingReferObject = obj.ReportingReferObject;
        this.Dimension = obj.Dimension;
        this.RefOrgId = obj.RefOrgId;
        this.RefOrgName = obj.RefOrgName;
        this.ReportingLineDefId = obj.ReportingLineDefId;
        this.ReportingLineDefName = obj.ReportingLineDefName;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ReportingLineParticipant(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const reportingLineParticipant = new ReportingLineParticipant(this.Model);
        reportingLineParticipant.Id = BpmnModelHelper.GenerateElementId();
        reportingLineParticipant.Name = this.Name;
        reportingLineParticipant.RefUserId = this.RefUserId;
        reportingLineParticipant.RefEmployeeId = this.RefEmployeeId;
        reportingLineParticipant.ReportingReferObject = this.ReportingReferObject;
        reportingLineParticipant.Dimension = this.Dimension;
        reportingLineParticipant.RefOrgId = this.RefOrgId;
        reportingLineParticipant.RefOrgName = this.RefOrgName;
        reportingLineParticipant.ReportingLineDefId = this.ReportingLineDefId;
        reportingLineParticipant.ReportingLineDefName = this.ReportingLineDefName;
        reportingLineParticipant.Tag = this.Tag;
        if (this.Condition) {
            reportingLineParticipant.Condition = this.Condition.clone();
        }
        return reportingLineParticipant;
    }
}
ReportingLineParticipant.Clr_Type_ID = 'wf.ReportingLineParticipant';
ReportingLineParticipant.CLR_TYPE_NAME = '汇报关系';
if (false) {
    /** @type {?} */
    ReportingLineParticipant.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    ReportingLineParticipant.CLR_TYPE_NAME;
    /** @type {?} */
    ReportingLineParticipant.prototype.ClrTypeID;
    /** @type {?} */
    ReportingLineParticipant.prototype.Tag;
    /** @type {?} */
    ReportingLineParticipant.prototype.ReportingLineDefId;
    /** @type {?} */
    ReportingLineParticipant.prototype.ReportingLineDefName;
    /** @type {?} */
    ReportingLineParticipant.prototype.RefOrgId;
    /** @type {?} */
    ReportingLineParticipant.prototype.RefOrgName;
    /** @type {?} */
    ReportingLineParticipant.prototype.RefUserId;
    /** @type {?} */
    ReportingLineParticipant.prototype.RefEmployeeId;
    /** @type {?} */
    ReportingLineParticipant.prototype.ReportingReferObject;
    /** @type {?} */
    ReportingLineParticipant.prototype.Dimension;
}
BpmnModelHelper.RegisterElementType(new ReportingLineParticipant(new BpmnModel()));
