// @gsp-wf/wf-process-model/esm2015/lib/model/participants/old/UserParticipant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { Participant } from "./Participant";
export class UserParticipant extends Participant {
    /**
     * @param {?} model
     * @param {?=} userId
     * @param {?=} name
     * @param {?=} orgType
     */
    constructor(model, userId, name, orgType) {
        super(model);
        this.ClrTypeID = UserParticipant.Clr_Type_ID;
        this.UserId = userId;
        this.Name = name;
        this.OrgType = orgType ? orgType : 'Sys';
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new UserParticipant(this.Model);
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "UserId", this.UserId);
        BpmnModelHelper.SetAttributeValue(obj, "OrgId", this.OrgId);
        BpmnModelHelper.SetAttributeValue(obj, "OrgName", this.OrgName);
        BpmnModelHelper.SetAttributeValue(obj, "OrgType", this.OrgType);
        BpmnModelHelper.SetAttributeValue(obj, "Tag", this.Tag);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.OrgId = obj.OrgId;
        this.OrgName = obj.OrgName;
        this.OrgType = obj.OrgType;
        this.UserId = obj.UserId;
        this.Tag = obj.Tag;
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const userParticipant = new UserParticipant(this.Model);
        userParticipant.Id = BpmnModelHelper.GenerateElementId();
        userParticipant.Name = this.Name;
        userParticipant.OrgId = this.OrgId;
        userParticipant.OrgName = this.OrgName;
        userParticipant.OrgType = this.OrgType;
        userParticipant.UserId = this.UserId;
        userParticipant.Tag = this.Tag;
        if (this.Condition) {
            userParticipant.Condition = this.Condition.clone();
        }
        return userParticipant;
    }
}
UserParticipant.Clr_Type_ID = "wf.UserParticipant";
UserParticipant.CLR_TYPE_NAME = '用户';
if (false) {
    /** @type {?} */
    UserParticipant.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    UserParticipant.CLR_TYPE_NAME;
    /** @type {?} */
    UserParticipant.prototype.ClrTypeID;
    /** @type {?} */
    UserParticipant.prototype.Tag;
    /** @type {?} */
    UserParticipant.prototype.UserId;
    /** @type {?} */
    UserParticipant.prototype.OrgId;
    /** @type {?} */
    UserParticipant.prototype.OrgName;
    /** @type {?} */
    UserParticipant.prototype.OrgType;
}
BpmnModelHelper.RegisterElementType(new UserParticipant(new BpmnModel()));
