// @farris/command-services/esm2015/lib/authority/authority.js
import { empty } from 'rxjs';
import { map } from 'rxjs/operators';
/**
 * 权限
 */
export class Authority {
    constructor() {
        this.operationLabelMap = new Map();
        this.authOfLabel = new Map();
    }
    /**
     * 设置label到operation id的映射
     * @param labelMap
     */
    setLabelMap(labelMap) {
        for (var i in labelMap) {
            this.operationLabelMap.set(i, labelMap[i]);
        }
        this.operationLabelMap.forEach((opId, label) => {
            this[label] = false;
        });
    }
    /**
     * 初始化权限的值
     */
    initialize(authorityService) {
        if (this.operationLabelMap && this.operationLabelMap.size > 0) {
            return authorityService.queryOperationAuthority(Array.from(this.operationLabelMap.values())).pipe(map((operationAuth) => {
                this.operationLabelMap.forEach((opId, label) => {
                    this.authOfLabel.set(label, operationAuth.has(opId) ? operationAuth.get(opId) : false);
                });
            }));
        }
        else {
            return empty();
        }
    }
    /**
     * 返回操作权限，即用户是否拥有Label所对应的权限
     * @param label 标签，可以是button id等
     */
    hasOperation(label) {
        return this.authOfLabel.has(label) && this.authOfLabel.get(label);
    }
}
