// @gsp-wf/wf-approval-logs/esm5/lib/wf-approval-logs.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { WfApprovalLogsComponent } from './wf-approval-logs.component';
import { TranslatePipe } from './service/translate.pipe';
import { CommonModule } from '@angular/common';
import { WfApprovalLogsService } from './wf-approval-logs.service';
import { FileListModule } from '@gsp-svc/file-viewer';
import { FormsModule } from '@angular/forms';
var WfApprovalLogsModule = /** @class */ (function () {
    function WfApprovalLogsModule() {
    }
    WfApprovalLogsModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        WfApprovalLogsComponent,
                        TranslatePipe,
                    ],
                    imports: [
                        CommonModule,
                        FileListModule,
                        FormsModule
                    ],
                    providers: [
                        WfApprovalLogsService,
                    ],
                    exports: [
                        WfApprovalLogsComponent
                    ],
                    entryComponents: [
                        WfApprovalLogsComponent,
                    ]
                },] }
    ];
    return WfApprovalLogsModule;
}());
export { WfApprovalLogsModule };
