// @gsp-wf/wf-approval-logs/esm5/lib/wf-approval-logs.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, ComponentFactoryResolver } from '@angular/core';
import { MessagerService } from '@farris/ui-messager';
import { BsModalService } from '@farris/ui-modal';
import { of } from 'rxjs';
import { translate } from './service/i18n/index';
import { WfApprovalLogsComponent } from './wf-approval-logs.component';
var WfApprovalLogsService = /** @class */ (function () {
    function WfApprovalLogsService(msgService, injector, resolver, modalService) {
        this.msgService = msgService;
        this.injector = injector;
        this.resolver = resolver;
        this.modalService = modalService;
    }
    /**
     * 根据流程实例id打开审批记录弹框
     * @param procInstId 流程实例id
     */
    /**
     * 根据流程实例id打开审批记录弹框
     * @param {?} procInstId 流程实例id
     * @return {?}
     */
    WfApprovalLogsService.prototype.showByProcInstId = /**
     * 根据流程实例id打开审批记录弹框
     * @param {?} procInstId 流程实例id
     * @return {?}
     */
    function (procInstId) {
        if (!procInstId) {
            console.error("查看审批记录，procInstId参数为空");
            return;
        }
        /** @type {?} */
        var parameters = {
            procInstId: procInstId
        };
        this.showApprovalLog(parameters);
    };
    /**
     * 根据表单数据id打开审批记录弹框
     * @param dataId 表单数据id
     * @param params 其他参数 params.style 为datagrid时，弹框使用大的（520,760,960）
     */
    /**
     * 根据表单数据id打开审批记录弹框
     * @param {?} dataId 表单数据id
     * @return {?}
     */
    WfApprovalLogsService.prototype.showByDataId = /**
     * 根据表单数据id打开审批记录弹框
     * @param {?} dataId 表单数据id
     * @return {?}
     */
    function (dataId) {
        if (!dataId) {
            console.error("查看审批记录，dataId参数为空");
            return;
        }
        /** @type {?} */
        var parameters = {
            dataId: dataId
        };
        this.showApprovalLog(parameters);
    };
    /**
     * 查看审批记录
     * @param parameters
     */
    /**
     * 查看审批记录
     * @param {?} parameters
     * @return {?}
     */
    WfApprovalLogsService.prototype.showApprovalLog = /**
     * 查看审批记录
     * @param {?} parameters
     * @return {?}
     */
    function (parameters) {
        /** @type {?} */
        var compFactory = this.resolver.resolveComponentFactory(WfApprovalLogsComponent);
        /** @type {?} */
        var compRef = compFactory.create(this.injector);
        compRef.instance.DataId = parameters.dataId;
        compRef.instance.Style = parameters.style;
        compRef.instance.ProcInstId = parameters.procInstId;
        compRef.instance.IncludeAdminOperate = parameters.includeAdminOperate;
        compRef.instance.IncludeHistory = parameters.includeHistory;
        compRef.instance.fill();
        /** @type {?} */
        var modalConfig = {
            title: this.getI18nValue('static.approvalLogs.title'),
            width: parameters.style == 'datagrid' ? 960 : 520,
            height: parameters.style == 'datagrid' ? 500 : 520,
            showButtons: parameters.style == 'datagrid' ? true : false,
            buttons: compRef.instance.taskDialogFooter,
            beforeClose: this.beforeClose,
        };
        this.dialog = this.modalService.show(compRef, modalConfig);
        this.addEventListener();
    };
    /**
     * 获取多语
     * @param name
     * @returns
     */
    /**
     * 获取多语
     * @private
     * @param {?} name
     * @return {?}
     */
    WfApprovalLogsService.prototype.getI18nValue = /**
     * 获取多语
     * @private
     * @param {?} name
     * @return {?}
     */
    function (name) {
        if (!name) {
            return '';
        }
        /** @type {?} */
        var defaultLang = localStorage.getItem('languageCode');
        /** @type {?} */
        var langData = defaultLang ? translate[defaultLang] : translate['zh-CHS'];
        /** @type {?} */
        var resultVal = '';
        if (name.indexOf('.') === -1) {
            resultVal = langData[name];
        }
        else {
            resultVal = name.split('.').reduce((/**
             * @param {?} obj
             * @param {?} key
             * @return {?}
             */
            function (obj, key) {
                if (obj) {
                    return obj[key];
                }
                else {
                    return null;
                }
            }), langData);
        }
        return resultVal;
    };
    //SPA模式下，点击附件，窗口关闭
    //SPA模式下，点击附件，窗口关闭
    /**
     * @private
     * @return {?}
     */
    WfApprovalLogsService.prototype.addEventListener = 
    //SPA模式下，点击附件，窗口关闭
    /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        window.addEventListener("message", (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            if (window.document.location.origin.search(event.origin) > -1) {
                if (event.data == 'TaskCloseLogDialog' && _this.dialog) {
                    _this.dialog.close();
                }
            }
        }), false);
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    WfApprovalLogsService.prototype.beforeClose = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        window.removeEventListener("message", (/**
         * @return {?}
         */
        function () { }));
        return of(true);
    };
    WfApprovalLogsService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    WfApprovalLogsService.ctorParameters = function () { return [
        { type: MessagerService },
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: BsModalService }
    ]; };
    return WfApprovalLogsService;
}());
export { WfApprovalLogsService };
if (false) {
    /** @type {?} */
    WfApprovalLogsService.prototype.dialog;
    /**
     * @type {?}
     * @private
     */
    WfApprovalLogsService.prototype.msgService;
    /**
     * @type {?}
     * @private
     */
    WfApprovalLogsService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    WfApprovalLogsService.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    WfApprovalLogsService.prototype.modalService;
}
