// @farris/ide-devkit/esm5/lib/ideframework/eventbus/duplex-event-pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { forkJoin, Observable, of, Subject, timer } from 'rxjs';
import { mapTo } from 'rxjs/operators';
var DuplexEventPipe = /** @class */ (function () {
    function DuplexEventPipe(eventName) {
        this.listenerResults = new Map();
        this.subject = new Subject();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    DuplexEventPipe.prototype.notify = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var thenFn = function () { };
        /** @type {?} */
        var catchFn = function () { };
        /** @type {?} */
        var timestamp = Date.now().toString();
        this.subject.next({ value: value, timestamp: timestamp });
        // fooObservable无意义，有两个作用：
        // 1. 没有实际observable时能够让forkJoin输出值，进入subscriber.next
        // 2. 建立延迟的observable（跳出当前micro task），避免subscription未返回和thenFn未被设置。
        /** @type {?} */
        var fooObservable = timer(0).pipe(mapTo(true));
        /** @type {?} */
        var listenerResultArray = this.listenerResults.get(timestamp);
        if (!listenerResultArray) {
            listenerResultArray = [fooObservable];
        }
        else {
            listenerResultArray = listenerResultArray.concat(fooObservable);
            this.listenerResults.delete(timestamp);
        }
        /** @type {?} */
        var subscription = forkJoin(listenerResultArray).subscribe({ next: function (results) {
                /** @type {?} */
                var success = results.reduce(function (previous, current) { return previous && current; });
                if (success) {
                    thenFn();
                }
                else {
                    catchFn();
                }
                subscription.unsubscribe();
            }, error: function () {
                subscription.unsubscribe();
            } });
        return {
            then: /**
             * @param {?} callback
             * @return {?}
             */
            function (callback) {
                thenFn = callback;
                return this;
            },
            catch: /**
             * @param {?} callback
             * @return {?}
             */
            function (callback) {
                catchFn = callback;
                return this;
            }
        };
    };
    /**
     * @param {?} callback
     * @param {?=} caller
     * @return {?}
     */
    DuplexEventPipe.prototype.listen = /**
     * @param {?} callback
     * @param {?=} caller
     * @return {?}
     */
    function (callback, caller) {
        var _this = this;
        /** @type {?} */
        var listener = caller ? callback.bind(caller) : callback;
        this.subject.subscribe(function (_a) {
            var value = _a.value, timestamp = _a.timestamp;
            /** @type {?} */
            var listenerResult;
            /** @type {?} */
            var invokeResult = listener(value);
            if (invokeResult instanceof Observable) {
                listenerResult = invokeResult;
            }
            else if (typeof invokeResult === 'boolean') {
                listenerResult = of(invokeResult);
            }
            else {
                listenerResult = of(true);
            }
            /** @type {?} */
            var listenerResultArray = _this.listenerResults.get(timestamp);
            if (!listenerResultArray) {
                listenerResultArray = [];
                _this.listenerResults.set(timestamp, listenerResultArray);
            }
            listenerResultArray.push(listenerResult);
        });
    };
    return DuplexEventPipe;
}());
export { DuplexEventPipe };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DuplexEventPipe.prototype.listenerResults;
    /**
     * @type {?}
     * @private
     */
    DuplexEventPipe.prototype.subject;
}
