// @farris/ide-devkit/esm5/lib/ideframework/eventbus/types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IDisposable() { }
if (false) {
    /**
     * @param {?} subscriber
     * @return {?}
     */
    IDisposable.prototype.dispose = function (subscriber) { };
}
/**
 * @record
 */
export function IEmitable() { }
if (false) {
    /**
     * @param {?} emitterType
     * @param {?} tokenValue
     * @param {?} eventName
     * @param {?} eventArgs
     * @return {?}
     */
    IEmitable.prototype.post = function (emitterType, tokenValue, eventName, eventArgs) { };
    /**
     * @param {?} target
     * @param {?} tokenValue
     * @param {?} eventName
     * @param {?} handler
     * @param {?} caller
     * @return {?}
     */
    IEmitable.prototype.on = function (target, tokenValue, eventName, handler, caller) { };
    /**
     * @param {?} target
     * @param {?} tokenValue
     * @param {?} eventName
     * @param {?} handler
     * @param {?} caller
     * @return {?}
     */
    IEmitable.prototype.once = function (target, tokenValue, eventName, handler, caller) { };
}
/**
 * @record
 */
export function IThenable() { }
if (false) {
    /**
     * @param {?} callback
     * @return {?}
     */
    IThenable.prototype.then = function (callback) { };
    /**
     * @param {?} callback
     * @return {?}
     */
    IThenable.prototype.catch = function (callback) { };
}
