// @gsp-wf/wf-process-editor/esm5/lib/views/approvel-items/entities/approval-item.entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ApprovalItemEntity = /** @class */ (function () {
    function ApprovalItemEntity(value, text, wfAction) {
        this.isDefault = false;
        this.commentNecessary = false;
        this.certification = false;
        this.attachmentNecessary = false;
        this.signatureNecessary = false;
        this.value = value;
        this.text = text;
        this.workFlowAction = wfAction;
    }
    return ApprovalItemEntity;
}());
export { ApprovalItemEntity };
if (false) {
    /** @type {?} */
    ApprovalItemEntity.prototype.value;
    /** @type {?} */
    ApprovalItemEntity.prototype.text;
    /** @type {?} */
    ApprovalItemEntity.prototype.workFlowAction;
    /** @type {?} */
    ApprovalItemEntity.prototype.isDefault;
    /** @type {?} */
    ApprovalItemEntity.prototype.commentNecessary;
    /** @type {?} */
    ApprovalItemEntity.prototype.certification;
    /** @type {?} */
    ApprovalItemEntity.prototype.attachmentNecessary;
    /** @type {?} */
    ApprovalItemEntity.prototype.signatureNecessary;
    /** @type {?} */
    ApprovalItemEntity.prototype.defaultComment;
}
