// @farris/mobile-devkit/esm2015/lib/changeset/types.js
/**
 * 变更记录
 */
export class Modification {
    /**
     * 构造函数
     * @param value 新值
     * @param modifyType 变更类型
     * @param path 变更路径
     * @param preValue 旧值
     */
    constructor(value, modifyType, path, preValue, position) {
        this.type = modifyType;
        this.value = value;
        this.preValue = preValue;
        this.path = path;
        this.position = position;
    }
}
/**
 * 变更类型
 */
export var ModifyType;
(function (ModifyType) {
    /**
     * 添加
     */
    ModifyType["Add"] = "ADD";
    /**
     * 添加数据
     */
    ModifyType["AddData"] = "AddData";
    /**
     * 克隆数据
     */
    ModifyType["Clone"] = "CLONE";
    /**
     * 删除
     */
    ModifyType["Remove"] = "REMOVE";
    ModifyType["RemoveData"] = "RemoveData";
    /**
     * 修改
     */
    ModifyType["ValueChange"] = "VALUE_CHANGE";
    /**
     * 加载
     */
    ModifyType["Load"] = "LOAD";
    /**
     * 未改变
     */
    ModifyType["UnChanged"] = "UNCHANGED";
    /**
     * 分页信息变更
     */
    ModifyType["PaginationInfoChange"] = "PAGINATION_INFO_CHANGE";
    /**
     * 插入
     */
    ModifyType["Insert"] = "Insert";
    /**
     * 更新实体
     */
    ModifyType["Update"] = "UPDATE";
})(ModifyType || (ModifyType = {}));
