// @farris/devkit/esm5/lib/expression/expression_result_factory.js
import { Injectable, Injector } from "@angular/core";
import { ExpressionManager } from "./expression_manager";
;
import { ExpressionRegistry } from "./expression_registry";
import { ExpressionResult } from "./expression_result";
var ExpressionResultFactory = /** @class */ (function () {
    function ExpressionResultFactory(injector, expressionRegistry, expressionManager, expressionResult) {
        this.injector = injector;
        this.expressionRegistry = expressionRegistry;
        this.expressionManager = expressionManager;
        this.expressionResult = expressionResult;
        this.registeEvent();
    }
    ExpressionResultFactory.prototype.registeEvent = function () {
        var _this = this;
        this.expressionRegistry.expressions.subscribe(function (expressionObjects) {
            // 加载完表达式之后做一次计算
            expressionObjects.forEach(function (expressionObject) {
                if (expressionObject.deps && expressionObject.deps.length > 0) {
                    return;
                }
                var result = _this.expressionManager.eval(expressionObject.id);
                _this.expressionResult[expressionObject.id] = result;
            });
        });
    };
    ExpressionResultFactory.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ExpressionResultFactory.ctorParameters = function () { return [
        { type: Injector },
        { type: ExpressionRegistry },
        { type: ExpressionManager },
        { type: ExpressionResult }
    ]; };
    return ExpressionResultFactory;
}());
export { ExpressionResultFactory };
