// @gsp-cmp/ccmp-ui/esm5/lib/ccmp-ui.dt-service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, EventEmitter, ComponentFactoryResolver } from '@angular/core';
import { CommonComponentSelectorComponent } from './ccmp-selector/common-component-selector.component';
import { ViewUtils } from './view-utils/view-utils';
import { BsModalService } from '@farris/ui-modal';
import { HttpHeaders, HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { ComponentInfo } from './params/component-info';
import * as i0 from "@angular/core";
import * as i1 from "./view-utils/view-utils";
import * as i2 from "@farris/ui-modal";
import * as i3 from "@angular/common/http";
var CommonComponentUiDTService = /** @class */ (function () {
    function CommonComponentUiDTService(resolver, injector, util, modalService, http) {
        this.resolver = resolver;
        this.injector = injector;
        this.util = util;
        this.modalService = modalService;
        this.http = http;
    }
    /**
     * 通用构件操作选择器
     * @param metadataId 通用构件元数据
     * @param sessionId 功能sessionId
     * @param options UI可选配置参数
     */
    /**
     * 通用构件操作选择器
     * @param {?} metadataId 通用构件元数据
     * @param {?} sessionId 功能sessionId
     * @param {?=} options UI可选配置参数
     * @return {?}
     */
    CommonComponentUiDTService.prototype.openSelector = /**
     * 通用构件操作选择器
     * @param {?} metadataId 通用构件元数据
     * @param {?} sessionId 功能sessionId
     * @param {?=} options UI可选配置参数
     * @return {?}
     */
    function (metadataId, sessionId, options) {
        var _this = this;
        return this.getComponentInfo(metadataId, sessionId).pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data.obselete) {
                _this.util.showObseleteDialog(data);
                return { confirmEvent: new EventEmitter(), closeEvent: new EventEmitter() };
            }
            else {
                return _this.create(data, options);
            }
        })));
    };
    /**
     * 获取构件信息
     * @param metadataId 构件元数据ID
     * @param sessionId 功能sessionid
     */
    /**
     * 获取构件信息
     * @param {?} metadataId 构件元数据ID
     * @param {?} sessionId 功能sessionid
     * @return {?}
     */
    CommonComponentUiDTService.prototype.getComponentInfo = /**
     * 获取构件信息
     * @param {?} metadataId 构件元数据ID
     * @param {?} sessionId 功能sessionid
     * @return {?}
     */
    function (metadataId, sessionId) {
        /** @type {?} */
        var headers = new HttpHeaders();
        headers.set("sessionId", sessionId);
        /** @type {?} */
        var url = "/api/dev/main/v1.0/common-component/component-info/" + metadataId;
        return this.http.get(url, { headers: headers }).pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data) {
                /** @type {?} */
                var cmpInfo = new ComponentInfo();
                cmpInfo.init(data);
                return cmpInfo;
            }
        })));
    };
    /**
     * @private
     * @param {?} componentInfo
     * @param {?} options
     * @return {?}
     */
    CommonComponentUiDTService.prototype.create = /**
     * @private
     * @param {?} componentInfo
     * @param {?} options
     * @return {?}
     */
    function (componentInfo, options) {
        /** @type {?} */
        var fac = this.resolver.resolveComponentFactory(CommonComponentSelectorComponent);
        /** @type {?} */
        var cmpRef = fac.create(this.injector);
        cmpRef.instance.componentInfo = componentInfo;
        /** @type {?} */
        var modalConfig = this.util.getModalConfig(cmpRef.instance.btnRef, options);
        /** @type {?} */
        var dialog = this.modalService.show(cmpRef, modalConfig);
        cmpRef.instance.dialog = dialog;
        return { confirmEvent: cmpRef.instance.confirmEmitter, closeEvent: cmpRef.instance.closeEmitter };
    };
    CommonComponentUiDTService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    CommonComponentUiDTService.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: Injector },
        { type: ViewUtils },
        { type: BsModalService },
        { type: HttpClient }
    ]; };
    /** @nocollapse */ CommonComponentUiDTService.ngInjectableDef = i0.defineInjectable({ factory: function CommonComponentUiDTService_Factory() { return new CommonComponentUiDTService(i0.inject(i0.ComponentFactoryResolver), i0.inject(i0.INJECTOR), i0.inject(i1.ViewUtils), i0.inject(i2.BsModalService), i0.inject(i3.HttpClient)); }, token: CommonComponentUiDTService, providedIn: "root" });
    return CommonComponentUiDTService;
}());
export { CommonComponentUiDTService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CommonComponentUiDTService.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    CommonComponentUiDTService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    CommonComponentUiDTService.prototype.util;
    /**
     * @type {?}
     * @private
     */
    CommonComponentUiDTService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    CommonComponentUiDTService.prototype.http;
}
