// @gsp-cmp/ccmp-ui/esm5/lib/ccmp-ui.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonComponentSelectorComponent } from './ccmp-selector/common-component-selector.component';
import { CommonComponentUiRTService } from './ccmp-ui.rt-service';
import { ViewUtils } from './view-utils/view-utils';
import { CommonModule } from '@angular/common';
import { ButtonsModule } from '@progress/kendo-angular-buttons';
import { GridModule } from '@progress/kendo-angular-grid';
import { HttpClientModule } from '@angular/common/http';
import { ModalModule } from '@farris/ui-modal';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { MessagerModule } from '@farris/ui-messager';
import { NotifyModule } from '@farris/ui-notify';
import { CommonComponentUiDTService } from './ccmp-ui.dt-service';
var CommonComponentUiModule = /** @class */ (function () {
    function CommonComponentUiModule() {
    }
    CommonComponentUiModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        CommonComponentSelectorComponent,
                    ],
                    imports: [
                        NotifyModule,
                        CommonModule,
                        ButtonsModule,
                        GridModule,
                        HttpClientModule,
                        ModalModule.forRoot(),
                        FarrisDialogModule,
                        MessagerModule.forRoot(),
                    ],
                    entryComponents: [CommonComponentSelectorComponent],
                    providers: [
                        CommonComponentUiRTService,
                        CommonComponentUiDTService,
                        ViewUtils
                    ],
                    exports: [CommonComponentSelectorComponent]
                },] }
    ];
    return CommonComponentUiModule;
}());
export { CommonComponentUiModule };
