// @gsp-sys/rtf-ui/fesm2015/gsp-sys-rtf-ui.js
import { CommonModule } from '@angular/common';
import { GridModule } from '@progress/kendo-angular-grid';
import { LayoutModule } from '@progress/kendo-angular-layout';
import { TreeViewModule } from '@progress/kendo-angular-treeview';
import { from } from 'rxjs';
import { Component, Output, EventEmitter, Input, Injectable, ViewChild, Injector, ComponentFactoryResolver, NgModule } from '@angular/core';
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { BsModalService, ModalModule } from '@farris/ui-modal';
import { MessagerService, MessagerModule } from '@farris/ui-messager';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class AppModuleTreeComponent {
    constructor() {
        /**
         * 菜单明细类
         */
        this.appTreeData = [];
        this.selectKeys = [];
        /**
         * 功能菜单模块变化后后
         */
        this.appTreeChanged = new EventEmitter();
        /**
         * 树节点展开后的输出
         */
        this.appTreeExpanded = new EventEmitter();
        /**
         * 树控件展开的节点值
         */
        this.expandedKeys = [];
    }
    /**
     * 变化情况
     * @param {?} changes 变化
     * @return {?}
     */
    ngOnChanges(changes) {
        // 初始化变化取消
        // 未包含appTreeData的变化
        if (!changes['appTreeData']) {
            return;
        }
        // 无效的值不予处理
        if (!changes['appTreeData'].currentValue ||
            changes['appTreeData'].currentValue.length === 0) {
            return;
        }
        // 获取到菜单数据
        from(this.appTreeData).subscribe((/**
         * @param {?} item
         * @return {?}
         */
        item => {
            if (!item.ParentId || item.ParentId === '0') {
                item.ParentId = undefined;
                if (item.Layer === 1) {
                    this.expandedKeys.push(item.Id);
                }
            }
        }));
        /*
         // 默认业务对象的处理
         if (!isNullOrUndefined(this.bizObjectId) && this.bizObjectId !== '') {
            // 树焦点行
            let curBizObjectId = this.bizObjectId;
            while (!isNullOrUndefined(curBizObjectId)) {
                this.expandedKeys.push(curBizObjectId);
                // 找到当前节点的付家店
                curBizObjectId = this.data.find(item => item['id'] === curBizObjectId)['parentID'];
            }
            // 右侧操作变化
            this.selectKeys = [this.bizObjectId];
            setTimeout(() => {
                this.bizObjectChanged.emit(this.bizObjectId);
            }, 80);
        }
        */
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * 选中行改变后触发事件
     * @param {?} args
     * @return {?}
     */
    selectChange(args) {
        /** @type {?} */
        const curApp = args.dataItem;
        // 首次触发
        if (this.selectKeys == null || this.selectKeys.length === 0) {
            this.appTreeChanged.emit(curApp);
        }
        else {
            // 这个事件里还不会对selectedkeys数组赋值，是上一个状态的
            /** @type {?} */
            const lastAppId = this.selectKeys[0];
            /** @type {?} */
            const curAppId = curApp.Id;
            // 变化的情况下触发
            if (lastAppId !== curAppId) {
                this.appTreeChanged.emit(curApp);
            }
        }
    }
    /**
     * @param {?} arg
     * @return {?}
     */
    onExpand(arg) {
        console.log('菜单帮助左侧树展开');
        if (!arg || !arg.dataItem) {
            return;
        }
        this.appTreeExpanded.emit(arg.dataItem);
    }
}
AppModuleTreeComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-module-tree',
                template: "<div class=\"d-flex flex-column flex-fill pt-2 pb-2\" style=\"height: 100%\">\r\n    <kendo-treeview [nodes]=\"appTreeData\" textField=\"Name\" kendoTreeViewExpandable kendoTreeViewFlatDataBinding\r\n        idField=\"Id\" parentIdField=\"ParentId\" kendoTreeViewSelectable [selectBy]=\"'Id'\"\r\n        [(selectedKeys)]=\"this.selectKeys\" [(expandedKeys)]=\"expandedKeys\" [expandBy]=\"'Id'\"\r\n        (expand)=\"this.onExpand($event)\" (selectionChange)=\"selectChange($event)\">\r\n    </kendo-treeview>\r\n</div>",
                styles: [""]
            }] }
];
/** @nocollapse */
AppModuleTreeComponent.ctorParameters = () => [];
AppModuleTreeComponent.propDecorators = {
    appTreeData: [{ type: Input }],
    bizObjectId: [{ type: Input }],
    appTreeChanged: [{ type: Output }],
    appTreeExpanded: [{ type: Output }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class RtfUIWebapiService {
    constructor() { }
    /**
     * 菜单api地址
     * @return {?}
     */
    get functionsApi() {
        return '/api/runtime/sys/v1.0/functions/allfuncs';
    }
    /**
     * 业务对象api地址
     * @return {?}
     */
    get bizObjectApi() {
        return '/api/dev/main/v1.0/business-objects';
    }
    /**
     * 分层获取菜单数据  api
     * @return {?}
     */
    get layerFunctionsApi() {
        return '/api/runtime/sys/v1.0/functions/layerFuncs';
    }
    /**
     * Http头信息
     * @return {?}
     */
    get httpHeader() {
        /** @type {?} */
        const options = {
            headers: {
                'Content-Type': 'application/json',
            },
            responseType: 'json'
        };
        return options;
    }
}
RtfUIWebapiService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RtfUIWebapiService.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class RtfAppService {
    /**
     * @param {?} rtfCommon
     * @param {?} httpClient
     */
    constructor(rtfCommon, httpClient) {
        this.rtfCommon = rtfCommon;
        this.httpClient = httpClient;
    }
    /**
     * 获取功能菜单操作
     * @return {?}
     */
    getAppList() {
        /** @type {?} */
        const url = this.rtfCommon.functionsApi;
        /** @type {?} */
        const header = this.rtfCommon.httpHeader;
        return this.httpClient.get(url, header);
    }
    /**
     * 根据祖父节点分层加载菜单数据
     * @param {?} grandParentId
     * @return {?}
     */
    getDataByGrandParentId(grandParentId) {
        /** @type {?} */
        const url = `${this.rtfCommon.layerFunctionsApi}/${grandParentId}`;
        /** @type {?} */
        const header = this.rtfCommon.httpHeader;
        return this.httpClient.get(url, header);
    }
}
RtfAppService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RtfAppService.ctorParameters = () => [
    { type: RtfUIWebapiService },
    { type: HttpClient }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const i18nFuncHelpInfo = {
    'zh-CHS': {
        TitleName: '功能菜单帮助',
        EnterNameOrCode: '输入编号或名称查找',
        Code: '编号',
        Name: '名称',
        Confirm: '确认',
        Cancel: '取消'
    },
    'zh-CHT': {
        TitleName: '功能菜單幫助',
        EnterNameOrCode: '輸入編號或名稱查找',
        Code: '編號',
        Name: ' 名稱',
        Confirm: '確認',
        Cancel: '取消'
    },
    en: {
        TitleName: 'Function Help',
        EnterNameOrCode: 'Please Enter Name Or Code',
        Code: 'Code',
        Name: 'Name',
        Confirm: 'Confirm',
        Cancel: 'Cancel'
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 系统管理：功能操作左树有、右列表展示
 */
class RtfAppComponent {
    /**
     * @param {?} rtfAppService
     */
    constructor(rtfAppService) {
        this.rtfAppService = rtfAppService;
        /**
         * 是否分层加载左侧树，默认为false
         */
        this.isLayeredLoad = false;
        /**
         * 确认事件
         */
        this.afterConfirm = new EventEmitter();
        /**
         * 关闭事件
         */
        this.afterCancle = new EventEmitter();
        /**
         * 功能操作全量数据
         */
        this.data = [];
        /**
         * 缓存的菜单及结构数据
         */
        this.allAppList = [];
        /**
         * 当前选中数据
         */
        this.curSelectedKeys = [];
        this.title = '功能菜单帮助';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.selectableSettings = {
            checkboxOnly: true,
            // 默认单选
            mode: !this.gridSelectMode ? 'single' : this.gridSelectMode,
            enabled: true
        };
        if (this.isLayeredLoad === false) {
            /** @type {?} */
            const allFuncsStream = this.rtfAppService.getAppList();
            this.loadTreeData(allFuncsStream);
        }
        else {
            this.loadTreeDataByGrandParentId('0');
        }
    }
    /**
     * @param {?} arg
     * @return {?}
     */
    onKeyUp(arg) {
        if (arg && arg.keyCode === 13) {
            this.searchData(this.filter.nativeElement.value);
        }
    }
    /**
     * 搜索
     * @param {?} filterValue 查询条件
     * @return {?}
     */
    searchData(filterValue) {
        // 空的搜索是全部数据
        if (filterValue === null || filterValue === '') {
            this.gridData = this.data.concat();
            return;
        }
        /** @type {?} */
        const filter = filterValue.toLowerCase();
        // 搜索到上级时
        this.gridData = this.data.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => item.Code.toLowerCase().includes(filter) ||
            item.Name.toLowerCase().includes(filter)));
    }
    /**
     * 业务对象变化后
     * @param {?} arg 业务对象参数
     * @return {?}
     */
    leftAppTreeChanged(arg) {
        if (!this.data || this.data.length === 0) {
            return;
        }
        /** @type {?} */
        const selectedId = arg.Id;
        /** @type {?} */
        const selectedPath = arg.Path;
        // 根据父子关系过滤当前界面数据，包含上下级
        this.gridData = this.data.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => item.ParentId === selectedId ||
            item.Path.startsWith(selectedPath)));
    }
    /**
     * @param {?} arg
     * @return {?}
     */
    onAppTreeExpanded(arg) {
        if (this.isLayeredLoad === false) {
            return;
        }
        if (this.isContainGrandChildNode(arg) === false) {
            this.loadTreeDataByGrandParentId(arg.Id);
        }
    }
    /**
     * 获取选中数据
     * @return {?}
     */
    getSelectedObject() {
        /** @type {?} */
        const result = this.data
            .filter((/**
         * @param {?} item
         * @return {?}
         */
        item => this.curSelectedKeys.findIndex((/**
         * @param {?} select
         * @return {?}
         */
        select => select === item.Id)) >= 0));
        return result;
    }
    /**
     * 确认
     * @return {?}
     */
    confirm() {
        /** @type {?} */
        const result = this.getSelectedObject();
        // this.selectedTreeNode.selectedInvok = result;
        this.afterConfirm.next(result);
        // 取消-目的是在这里让模态框消失
        this.cancle();
    }
    /**
     * 取消
     * @return {?}
     */
    cancle() {
        this.afterCancle.next();
    }
    /**
     * @param {?} langCode
     * @return {?}
     */
    setCurrentLangCode(langCode) {
        this.currentLangCode = langCode;
        this.title = this.getResOnCurrentLangeContext('TitleName');
    }
    /**
     * @param {?} resKey
     * @return {?}
     */
    getResOnCurrentLangeContext(resKey) {
        /** @type {?} */
        const resValue = i18nFuncHelpInfo[this.currentLangCode][resKey];
        return resValue;
    }
    /**
     * @private
     * @param {?} node
     * @return {?}
     */
    isContainGrandChildNode(node) {
        if (!('1' === node.Layer || '2' === node.Layer)) {
            return true;
        }
        /** @type {?} */
        const parentId = node.Id;
        /** @type {?} */
        const sonList = this.allAppList.filter((/**
         * @param {?} f
         * @return {?}
         */
        f => f.FParentid === parentId));
        /** @type {?} */
        let res = false;
        for (let i = 0; i < sonList.length; i++) {
            /** @type {?} */
            const grandList = this.allAppList.filter((/**
             * @param {?} o
             * @return {?}
             */
            o => o.FParentid === sonList[i].FID));
            if (grandList && grandList.length > 0) {
                res = true;
                break;
            }
        }
        return res;
    }
    /**
     * @private
     * @param {?} grandParentId
     * @return {?}
     */
    loadTreeDataByGrandParentId(grandParentId) {
        // 分层加载
        /** @type {?} */
        const stream = this.rtfAppService.getDataByGrandParentId(grandParentId);
        this.loadTreeData(stream);
    }
    /**
     * @private
     * @param {?} stream
     * @return {?}
     */
    loadTreeData(stream) {
        stream.subscribe((/**
         * @param {?} res
         * @return {?}
         */
        (res) => {
            if (res) {
                res.forEach((/**
                 * @param {?} element
                 * @return {?}
                 */
                element => {
                    /** @type {?} */
                    const temp = {
                        Id: element.id,
                        Code: element.code,
                        Name: element.name,
                        ParentId: element.parentId,
                        Layer: element.layer,
                        Path: element.path,
                        SortOrder: element.sortOrder,
                        Child: element.child,
                        InvokingConfig: element.invokingConfig
                    };
                    this.allAppList.push(temp);
                }));
            }
            // 左侧树
            this.treeData = this.allAppList.filter((/**
             * @param {?} item
             * @return {?}
             */
            item => item.Layer <= 3));
            // 菜单明细数据
            this.data = this.allAppList.filter((/**
             * @param {?} item
             * @return {?}
             */
            item => item.Layer >= 4));
        }), (/**
         * @param {?} err
         * @return {?}
         */
        (err) => {
            console.log('获取菜单数据失败');
        }));
    }
}
RtfAppComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-rtf-app',
                template: "<div class=\"d-flex flex-row\" style=\"width: 800px; border:2px;position:absolute;top:0px;bottom: 0px;\r\nright:0px;left:0px;overflow: hidden;height: 100%;\">\r\n    <div class=\"flex-fill\">\r\n        <kendo-splitter style=\"height:100%\" orientation=\"horizontal\">\r\n            <kendo-splitter-pane size=\"40%\" min=\"20%\" max=\"50%\">\r\n                <div class=\"d-flex flex-column flex-fill\" style=\"height:100%\">\r\n                    <div class=\"p-2\" style=\"height: 35px;\">\r\n                        <h5>{{this.getResOnCurrentLangeContext('TitleName')}}</h5>\r\n                        <hr class=\"m-0 p-0\" />\r\n                    </div>\r\n                    <div class=\"flex-fill\" style=\"overflow: auto;\">\r\n                        <app-module-tree [appTreeData]=\"this.treeData\" [bizObjectId]=\"bizObjectId\"\r\n                            (appTreeChanged)=\"leftAppTreeChanged($event)\" (appTreeExpanded)=\"onAppTreeExpanded($event)\">\r\n                        </app-module-tree>\r\n                    </div>\r\n                </div>\r\n            </kendo-splitter-pane>\r\n            <kendo-splitter-pane>\r\n                <div class=\"d-flex flex-column flex-fill\" style=\"height:100%\">\r\n                    <!-- <div class=\"p-2\" style=\"height: 35px;\">\r\n                        <h5>\u83DC\u5355\u5217\u8868</h5>\r\n                        <hr class=\"m-0 p-0\" />\r\n                    </div> -->\r\n                    <div class=\"clearfix mt-1 mb-1 ml-1\">\r\n                        <div class=\"f-cmp-inputgroup input-group\" style=\"width: 70%;\">\r\n                            <input #filter class=\"form-control\" (keyup)=\"onKeyUp($event)\"\r\n                                placeholder=\"{{this.getResOnCurrentLangeContext('EnterNameOrCode')}}\">\r\n                            <div class=\"input-group-append\"><span class=\"input-group-text\"\r\n                                    (click)=\"searchData(filter.value)\">\r\n                                    <i class=\"k-icon k-i-search\"></i></span></div>\r\n                        </div>\r\n                    </div>\r\n                    <div class=\"d-flex flex-fill\" style=\"position:relative;\">\r\n                        <kendo-grid [class]=\"'flex-fill'\" [selectable]=\"selectableSettings\"\r\n                            [kendoGridBinding]=\"gridData\" [kendoGridSelectBy]=\"'Id'\" [selectedKeys]=\"curSelectedKeys\">\r\n                            <kendo-grid-checkbox-column width=\"50px\"></kendo-grid-checkbox-column>\r\n                            <kendo-grid-column field=\"Code\" title=\"{{this.getResOnCurrentLangeContext('Code')}}\">\r\n                            </kendo-grid-column>\r\n                            <kendo-grid-column field=\"Name\" title=\"{{this.getResOnCurrentLangeContext('Name')}}\">\r\n                            </kendo-grid-column>\r\n                        </kendo-grid>\r\n                    </div>\r\n                </div>\r\n            </kendo-splitter-pane>\r\n        </kendo-splitter>\r\n    </div>\r\n</div>\r\n<ng-template #btnOkAndCancle>\r\n    <button class=\"btn btn-sm btn-primary\" (click)=\"confirm()\">{{this.getResOnCurrentLangeContext('Confirm')}}</button>\r\n    <button class=\"btn btn-sm btn-default\" (click)=\"cancle()\">{{this.getResOnCurrentLangeContext('Cancel')}}</button>\r\n</ng-template>",
                styles: [""]
            }] }
];
/** @nocollapse */
RtfAppComponent.ctorParameters = () => [
    { type: RtfAppService }
];
RtfAppComponent.propDecorators = {
    bizObjectId: [{ type: Input }],
    gridSelectMode: [{ type: Input }],
    afterConfirm: [{ type: Output }],
    afterCancle: [{ type: Output }],
    btnOkAndCancleRef: [{ type: ViewChild, args: ['btnOkAndCancle',] }],
    filter: [{ type: ViewChild, args: ['filter',] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// tslint:disable
class RtfAppHelpComponent {
    /**
     * @param {?} rtfCommon
     * @param {?} modalService
     * @param {?} farrisMsg
     * @param {?} injector
     * @param {?} cfr
     */
    constructor(rtfCommon, modalService, farrisMsg, injector, cfr) {
        this.rtfCommon = rtfCommon;
        this.modalService = modalService;
        this.farrisMsg = farrisMsg;
        this.injector = injector;
        this.cfr = cfr;
        /**
         * 初始选中的权限对象内码
         */
        this.initSelectKeys = [];
        this.languageCode = 'zh-CHS';
        this.isLayeredLoad = false;
        /**
         * 输出事件
         */
        this.afterHelpConfirm = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * 弹出Farris模态框帮助
     * @param {?=} sessionId
     * @return {?}
     */
    showHelp(sessionId) {
        // if (!sessionId || sessionId === '') {
        //     this.farrisMsg.error('调用帮助时请传递上下文变量!');
        //     return;
        // }
        this.rtfCommon.sessionId = ' ';
        /** @type {?} */
        const compFactory = this.cfr.resolveComponentFactory(RtfAppComponent);
        this.funcOpComp = compFactory.create(this.injector);
        // 向组件绑定语言代码
        this.funcOpComp.instance.setCurrentLangCode(this.languageCode);
        this.funcOpComp.instance.isLayeredLoad = this.isLayeredLoad;
        // 获取功能菜单的实例组件
        if (this.initSelectKeys != null && this.initSelectKeys.length > 0) {
            this.initSelectKeys
                .forEach((/**
             * @param {?} key
             * @return {?}
             */
            key => this.funcOpComp.instance.curSelectedKeys.push(key)));
        }
        // 初始化选择的业务对象
        if (this.bizObjectId && this.bizObjectId !== '') {
            this.funcOpComp.instance.bizObjectId = this.bizObjectId;
        }
        /** @type {?} */
        const modalConfig = {
            title: this.funcOpComp.instance.title,
            width: 800,
            height: 550,
            buttons: this.funcOpComp.instance.btnOkAndCancleRef,
            showButtons: true,
            showMaxButton: false,
            resizable: false
        };
        // 弹出帮助框
        this.funcOpDialog = this.modalService.show(this.funcOpComp, modalConfig);
        // 取消后
        this.funcOpComp.instance.afterCancle.subscribe((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            // 消失模态框
            this.funcOpDialog.close();
        }));
        // 确认后
        this.funcOpComp.instance.afterConfirm.subscribe((/**
         * @param {?} res
         * @return {?}
         */
        (res) => {
            this.afterHelpConfirm.next(res);
        }));
    }
}
RtfAppHelpComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-rtf-app-help',
                template: "",
                styles: [""]
            }] }
];
/** @nocollapse */
RtfAppHelpComponent.ctorParameters = () => [
    { type: RtfUIWebapiService },
    { type: BsModalService },
    { type: MessagerService },
    { type: Injector },
    { type: ComponentFactoryResolver }
];
RtfAppHelpComponent.propDecorators = {
    bizObjectId: [{ type: Input }],
    initSelectKeys: [{ type: Input }],
    languageCode: [{ type: Input }],
    isLayeredLoad: [{ type: Input }],
    afterHelpConfirm: [{ type: Output }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class RtfAppHelpModule {
}
RtfAppHelpModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    HttpClientModule,
                    // kendo
                    GridModule,
                    LayoutModule,
                    TreeViewModule,
                    // farris
                    ModalModule.forRoot(),
                    MessagerModule.forRoot(),
                ],
                declarations: [
                    AppModuleTreeComponent,
                    RtfAppComponent,
                    RtfAppHelpComponent,
                ],
                entryComponents: [
                    RtfAppComponent
                ],
                providers: [
                    RtfUIWebapiService,
                    RtfAppService,
                ],
                exports: [
                    RtfAppComponent,
                    RtfAppHelpComponent
                ]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { RtfAppHelpModule, RtfAppHelpComponent, RtfAppComponent, RtfUIWebapiService as ɵc, AppModuleTreeComponent as ɵa, RtfAppService as ɵb };
