// @ecp-caf/common-structure/esm2015/lib/serialize-context/default/default-serialize-context.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DefaultDataTypeSerializerFactory } from './factory/default-data-type-serializer-factory';
import { DefaultOperationSerializerFactory } from './factory/default-operation-serializer-factory';
import { DefaultParameterSerializerFactory } from './factory/default-parameter-serializer-factory';
import { DefaultPropertySerializerFactory } from './factory/default-property-serializer-factory';
import { DefaultStructureSerializerFactory } from './factory/default-structure-serializer-factory';
import { DefaultConstraintSerializerFactory } from './factory/default-constraint-serializer-factory';
export class DefaultSerializeContext {
    /**
     * @return {?}
     */
    getDataTypeSerializerFactory() {
        if (!this.dataTypeFactory) {
            this.dataTypeFactory = new DefaultDataTypeSerializerFactory();
        }
        return this.dataTypeFactory;
    }
    /**
     * @return {?}
     */
    getOperationSerializerFactory() {
        if (!this.operationFactory) {
            this.operationFactory = new DefaultOperationSerializerFactory();
        }
        return this.operationFactory;
    }
    /**
     * @return {?}
     */
    getParameterSerializerFactory() {
        if (!this.parameterFactory) {
            this.parameterFactory = new DefaultParameterSerializerFactory();
        }
        return this.parameterFactory;
    }
    /**
     * @return {?}
     */
    getPropertySerializerFactory() {
        if (!this.propertyFactory) {
            this.propertyFactory = new DefaultPropertySerializerFactory();
        }
        return this.propertyFactory;
    }
    /**
     * @return {?}
     */
    getStructureSerializerFactory() {
        if (!this.structureFactory) {
            this.structureFactory = new DefaultStructureSerializerFactory();
        }
        return this.structureFactory;
    }
    /**
     * @return {?}
     */
    getConstraintSerializerFactory() {
        if (!this.constraintFactory) {
            this.constraintFactory = new DefaultConstraintSerializerFactory();
        }
        return this.constraintFactory;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    DefaultSerializeContext.prototype.dataTypeFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultSerializeContext.prototype.operationFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultSerializeContext.prototype.parameterFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultSerializeContext.prototype.propertyFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultSerializeContext.prototype.structureFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultSerializeContext.prototype.constraintFactory;
}
