// @ecp-caf/common-structure/esm2015/lib/serialize-context/default/default-constant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * DefaultDataType
 * @type {?}
 */
export const DEFAULT_PRIMITIVE_TYPE = "DefaultPrimitiveType";
/** @type {?} */
export const DEFAULT_COLLECTION_TYPE = "DefaultCollectionType";
/** @type {?} */
export const DEFAULT_MAP_TYPE = "DefaultMapType";
/** @type {?} */
export const DEFAULT_STRUCTURED_TYPE = "DefaultStructuredType";
/** @type {?} */
export const DEFAULT_DATA_TRANSFER_OBJECT_TYPE = "DefaultDtoType";
/** @type {?} */
export const DEFAULT_ENTITY_TYPE = "DefaultEntityType";
/** @type {?} */
export const DEFAULT_VALUE_OBJECT_TYPE = "DefaultValueObjectType";
/** @type {?} */
export const DEFAULT_ENUMERATION_TYPE = "DefaultEnumType";
/** @type {?} */
export const DEFAULT_OBJECT_TYPE = "Object";
/**
 * DefaultElement
 * @type {?}
 */
export const DEFAULT_PROPERTY = "DefaultProperty";
/** @type {?} */
export const DEFAULT_PARAMETER = "DefaultParameter";
/** @type {?} */
export const DEFAULT_OPERATION = "DefaultOperation";
/** @type {?} */
export const DEFAULT_MUTILLINGUAL_PROPERTY = "DefaultMultillingualProperty";
/**
 * DefaultStructure
 * @type {?}
 */
export const DEFAULT_ENTITY_OBJECT = "DefaultEntityObject";
/** @type {?} */
export const DEFAULT_DATA_TRANSFER_OBJECT = "DefaultDataTransferObject";
/** @type {?} */
export const DEFAULT_VALUE_OBJECT = "DefaultValueObject";
/**
 * DefaultConstraint
 * @type {?}
 */
export const DEFAULT_CONSTRAINT = "DefaultConstraint";
