// @edp-pmf/grapheditor/dist/esm/lib/util/MxGraphInnerClass.js
/**
 * mxgraph内置类，并未导出
 */
export var MxGraphInnerClass;
(function (MxGraphInnerClass) {
    MxGraphInnerClass["mxCellCodec"] = "mxCellCodec";
    MxGraphInnerClass["mxChildChangeCodec"] = "mxChildChangeCodec";
    MxGraphInnerClass["mxDefaultKeyHandlerCodec"] = "mxDefaultKeyHandlerCodec";
    MxGraphInnerClass["mxDefaultPopupMenuCodec"] = "mxDefaultPopupMenuCodec";
    MxGraphInnerClass["mxEditorCodec"] = "mxEditorCodec";
    MxGraphInnerClass["mxGraphCodec"] = "mxGraphCodec";
    MxGraphInnerClass["mxGraphViewCodec"] = "mxGraphViewCodec";
    MxGraphInnerClass["mxModelCodec"] = "mxModelCodec";
    MxGraphInnerClass["mxRootChangeCodec"] = "mxRootChangeCodec";
    MxGraphInnerClass["mxTerminalChangeCodec"] = "mxTerminalChangeCodec";
    MxGraphInnerClass["mxRootChange"] = "mxRootChange";
    MxGraphInnerClass["mxChildChange"] = "mxChildChange";
    MxGraphInnerClass["mxTerminalChange"] = "mxTerminalChange";
    MxGraphInnerClass["mxValueChange"] = "mxValueChange";
    MxGraphInnerClass["mxStyleChange"] = "mxStyleChange";
    MxGraphInnerClass["mxGeometryChange"] = "mxGeometryChange";
    MxGraphInnerClass["mxCollapseChange"] = "mxCollapseChange";
    MxGraphInnerClass["mxVisibleChange"] = "mxVisibleChange";
    MxGraphInnerClass["mxCellAttributeChange"] = "mxCellAttributeChange";
    MxGraphInnerClass["renderingHint"] = "renderingHint";
    MxGraphInnerClass["mxSelectionChange"] = "mxSelectionChange";
    MxGraphInnerClass["mxCurrentRootChange"] = "mxCurrentRootChange";
})(MxGraphInnerClass || (MxGraphInnerClass = {}));
