// @gsp-sys/rac-common/esm2015/lib/common/util/rac-util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpHeaders } from '@angular/common/http';
import { SessionService } from '@ecp-caf/caf-common';
import { BizDataDTO } from '../entity/auresult.entity';
export class RacCommonUtil {
    /**
     * @param {?} sessionSv
     */
    constructor(sessionSv) {
        this.sessionSv = sessionSv;
        /**
         * 维度1的值
         */
        this.assid1 = '';
        /**
         * 维度2的值
         */
        this.assid2 = '';
        this.AUDATACONTROL = 'AuDataControl';
        this.expandIcon = 'f-icon f-icon-file-folder-open text-warning';
        this.collapseIcon = 'f-icon f-icon-file-folder-close text-warning';
        this.leafIcon = 'f-icon f-icon-file-txt text-warning';
    }
    /**
     * 构造Header
     * @return {?}
     */
    getHeader() {
        /** @type {?} */
        const sessionId = this.sessionSv.getCurrentSeesionId();
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', sessionId);
        return { headers };
    }
    /**
     * 获取上下文中的用户
     * @return {?}
     */
    getCurrentUserId() {
        // 测试用写死
        return 'admin';
    }
    /**
     * Guid生成
     * @return {?}
     */
    getGuid36() {
        /**
         * @return {?}
         */
        function S4() {
            // tslint:disable-next-line:no-bitwise
            return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
        }
        return (S4() + S4() + '-' + S4() + '-' + S4() + '-' + S4() + '-' + S4() + S4() + S4());
    }
    /**
     * 权限分配结果路径
     * @return {?}
     */
    getResultPath() {
        /** @type {?} */
        const path = `/api/runtime/sys/v1.0/assignresults`;
        return path;
    }
    /**
     * 权限对象路径
     * @return {?}
     */
    getAuthorizationPath() {
        /** @type {?} */
        const path = `/api/runtime/sys/v1.0/authorizations`;
        return path;
    }
    /**
     * 权限字段路径
     * @return {?}
     */
    getAuthfieldPath() {
        /** @type {?} */
        const path = `/api/runtime/sys/v1.0/authfields`;
        return path;
    }
    /**
     * 重授权对象路径
     * @return {?}
     */
    getReAssignPath() {
        /** @type {?} */
        const path = `/api/runtime/sys/v1.0/reassigns`;
        return path;
    }
    /**
     * 获取新增的内容
     * @param {?} newarr
     * @param {?} oldarr
     * @return {?}
     */
    getAddIds(newarr, oldarr) {
        /** @type {?} */
        const addIds = [];
        newarr.forEach((/**
         * @param {?} x
         * @return {?}
         */
        x => {
            if (!oldarr.includes(x)) {
                addIds.push(x);
            }
        }));
        return addIds;
    }
    /**
     * 获取删除的内容
     * @param {?} newarr
     * @param {?} oldarr
     * @return {?}
     */
    getRemoveIds(newarr, oldarr) {
        return this.getAddIds(oldarr, newarr);
    }
    /**
     * @param {?} arrs
     * @return {?}
     */
    convertToBizDataDTO(arrs) {
        return arrs.map((/**
         * @param {?} x
         * @return {?}
         */
        x => {
            /** @type {?} */
            const dataDto = new BizDataDTO();
            dataDto.data = x;
            return dataDto;
        }));
    }
    /**
     * 有效的字符串
     * @param {?} value
     * @return {?}
     */
    isNullOrEmptyString(value) {
        return this.isBlank(value) || value.toString().trim().length === 0;
    }
    /**
     * 不为空
     * @param {?} value
     * @return {?}
     */
    isBlank(value) {
        return value === null || value === undefined;
    }
    /**
     * 字符串不为空
     * @param {?} value
     * @return {?}
     */
    isNotNullOrEmptyString(value) {
        return !(this.isNullOrEmptyString(value));
    }
    /**
     * 找到树形Parent分级方式的加载的第一层
     * @param {?} datas
     * @param {?} key
     * @param {?} parentKey
     * @return {?}
     */
    findLayer1ByParentId(datas, key, parentKey) {
        /** @type {?} */
        const rtn = [];
        datas.forEach((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            /** @type {?} */
            const findParent = datas.find((/**
             * @param {?} parent
             * @return {?}
             */
            parent => parent[key] === data[parentKey]));
            if (!findParent && data.type !== undefined && data.type !== null && data.type.toString() === '99') {
                data.parentId = undefined;
                rtn.push(data);
            }
        }));
        return rtn;
    }
    /**
     * 类型是否为导航
     * @param {?} type
     * @return {?}
     */
    isNav(type) {
        return type === '99' || type === '0';
    }
    /**
     * @param {?} farrisTt
     * @return {?}
     */
    setTreeIcon(farrisTt) {
        farrisTt.expandIcon = this.expandIcon;
        farrisTt.collapseIcon = this.collapseIcon;
        farrisTt.leafIcon = this.leafIcon;
        farrisTt.detectChanges();
    }
}
RacCommonUtil.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RacCommonUtil.ctorParameters = () => [
    { type: SessionService }
];
if (false) {
    /** @type {?} */
    RacCommonUtil.prototype.sessionId;
    /** @type {?} */
    RacCommonUtil.prototype.ownerId;
    /** @type {?} */
    RacCommonUtil.prototype.ownerCode;
    /** @type {?} */
    RacCommonUtil.prototype.ownerName;
    /**
     * 权限所有者类型 岗位1，用户2
     * @type {?}
     */
    RacCommonUtil.prototype.ownerType;
    /**
     * 岗位类型 组织岗0，通用岗1，用户2 TODO 命令不支持传参数，暂时这么处理
     * @type {?}
     */
    RacCommonUtil.prototype.posType;
    /** @type {?} */
    RacCommonUtil.prototype.packageId;
    /** @type {?} */
    RacCommonUtil.prototype.year;
    /**
     * 维度1的值
     * @type {?}
     */
    RacCommonUtil.prototype.assid1;
    /**
     * 维度2的值
     * @type {?}
     */
    RacCommonUtil.prototype.assid2;
    /** @type {?} */
    RacCommonUtil.prototype.funcId;
    /** @type {?} */
    RacCommonUtil.prototype.AUDATACONTROL;
    /** @type {?} */
    RacCommonUtil.prototype.expandIcon;
    /** @type {?} */
    RacCommonUtil.prototype.collapseIcon;
    /** @type {?} */
    RacCommonUtil.prototype.leafIcon;
    /**
     * @type {?}
     * @private
     */
    RacCommonUtil.prototype.sessionSv;
}
