// @gsp-wf/wf-process-editor/esm5/lib/designer/wf-cell/AifCreationBillActivityCell.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { NodeCell, ElementPropertyConfig } from "@edp-pmf/grapheditor";
import { mxResources } from "@edp-pmf/mxgraph-ts";
import { AifCreationBillActivity } from "@gsp-wf/wf-process-model";
import { BpmnModelHelper, FlowNodeState } from "@edp-pmf/bpmn-model";
import { WfConstants } from "../WfConstants";
import { AifCreationBillActivityPropertyKeys } from "./property-keys/AifCreationBillActivityPropertyKeys";
/**
 * 自动生单活动节点
 */
var AifCreationBillActivityCell = /** @class */ (function (_super) {
    tslib_1.__extends(AifCreationBillActivityCell, _super);
    function AifCreationBillActivityCell(value, geometry, style) {
        var _a;
        var _this = _super.call(this, value, geometry, style) || this;
        _this.clrTypeId = AifCreationBillActivityCell.CLR_TYPE_ID;
        _this.name = "自动生单";
        _this.nameLanguage = {
            "zh-CHS": "自动生单",
            "zh-CHT": "自動生單",
            "en": "Creation Bill ",
        };
        _this.imgUrls = (_a = {},
            _a[FlowNodeState.Default] = WfConstants.IMAGE_PATH + "/aifCreationActivity.png",
            _a[FlowNodeState.Selected] = WfConstants.IMAGE_PATH + "/aifCreationActivity_select.svg",
            _a);
        return _this;
    }
    /**
     * @param {?=} flowElt
     * @return {?}
     */
    AifCreationBillActivityCell.prototype.createFlowElement = /**
     * @param {?=} flowElt
     * @return {?}
     */
    function (flowElt) {
        /** @type {?} */
        var flowNode = (/** @type {?} */ (_super.prototype.createFlowElement.call(this, flowElt)));
        flowNode.Id = "AifCreationBillActivity" + BpmnModelHelper.GenerateElementId();
        return flowNode;
    };
    /**
     * @return {?}
     */
    AifCreationBillActivityCell.prototype.getFarrisPropConfig = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var act = this.flowElement;
        /** @type {?} */
        var PropertyConfig = new ElementPropertyConfig();
        /** @type {?} */
        var propertyData = new Object();
        PropertyConfig.categoryId = 'general';
        PropertyConfig.categoryName = mxResources.get('basicInformation');
        PropertyConfig.tabId = 'basic';
        PropertyConfig.tabName = mxResources.get('basicAttributes');
        PropertyConfig.properties = [
            { propertyID: AifCreationBillActivityPropertyKeys.Name, propertyName: mxResources.get('name'), propertyType: 'string', visible: !this.editorUi.multiLangEnabled },
            { propertyID: AifCreationBillActivityPropertyKeys.NameLanguage, propertyName: mxResources.get('nameLanguage'), propertyType: 'multiLanguage', visible: this.editorUi.multiLangEnabled },
            { propertyID: AifCreationBillActivityPropertyKeys.Id, propertyName: mxResources.get('id'), propertyType: 'string', readonly: true }
        ];
        /** @type {?} */
        var bifDef = new ElementPropertyConfig();
        bifDef.categoryId = 'bifDef';
        bifDef.categoryName = mxResources.get('bifDef');
        bifDef.tabId = 'basic';
        bifDef.tabName = mxResources.get('basicAttributes');
        bifDef.properties = [
            { propertyID: AifCreationBillActivityPropertyKeys.aifProcessCode, propertyName: mxResources.get('aifProcess'), propertyType: 'modal' },
        ];
        propertyData[AifCreationBillActivityPropertyKeys.Id] = act.Id;
        propertyData[AifCreationBillActivityPropertyKeys.ClrTypeID] = act.ClrTypeID;
        propertyData[AifCreationBillActivityPropertyKeys.Name] = act.Name || mxResources.get('autoAif');
        propertyData[AifCreationBillActivityPropertyKeys.NameLanguage] = act.NameLanguage || {
            "zh-CHS": act.Name,
        };
        propertyData[AifCreationBillActivityPropertyKeys.aifProcessCode] = act.aifProcessCode;
        propertyData[AifCreationBillActivityPropertyKeys.isRelied] = act.isRelied;
        return {
            propertyConfig: [PropertyConfig, bifDef],
            propertyData: propertyData
        };
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AifCreationBillActivityCell.prototype.updateProps = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var propertyId = obj.propertyID;
        switch (propertyId) {
            case AifCreationBillActivityPropertyKeys.aifProcessCode:
                this.resetProcessCode(this.flowElement, obj);
                break;
            default:
                _super.prototype.updateProps.call(this, obj);
                break;
        }
    };
    /**
     * @param {?} startActivity
     * @param {?} obj
     * @return {?}
     */
    AifCreationBillActivityCell.prototype.resetProcessCode = /**
     * @param {?} startActivity
     * @param {?} obj
     * @return {?}
     */
    function (startActivity, obj) {
        /** @type {?} */
        var data = obj.propertyValue;
        if (data) {
            startActivity.aifProcessCode = data;
        }
        else {
            startActivity.aifProcessCode = null;
        }
    };
    AifCreationBillActivityCell.CLR_TYPE_ID = AifCreationBillActivity.Clr_Type_ID;
    return AifCreationBillActivityCell;
}(NodeCell));
export { AifCreationBillActivityCell };
if (false) {
    /** @type {?} */
    AifCreationBillActivityCell.CLR_TYPE_ID;
    /** @type {?} */
    AifCreationBillActivityCell.prototype.clrTypeId;
    /** @type {?} */
    AifCreationBillActivityCell.prototype.name;
    /** @type {?} */
    AifCreationBillActivityCell.prototype.nameLanguage;
    /** @type {?} */
    AifCreationBillActivityCell.prototype.flowElement;
    /** @type {?} */
    AifCreationBillActivityCell.prototype.imgUrls;
}
