// @gsp-wf/wf-process-editor/esm5/lib/designer/wf-cell/WfProcessCell.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ElementPropertyConfig, PmfEvents, ProcessCell } from "@edp-pmf/grapheditor";
import { BpmnModelHelper, DataTypes, DataType, DataObject } from "@edp-pmf/bpmn-model";
import { ProcessHeader, StartMode, Component, CDTSchemaType, ToolTime, ExecuteForm, VariableSetting, ActualParameter, TitleExpression, CustomizeAction, MessageNotice, TimeLimitControl, ExtendProperty, ImageSettings, PositionParticipantExpression, UserParticipantExpression, ValuedDataObject, TimeLimitControlInfo, SendMessageExpireOperationInfo, SendMessageExpireOperation, AutoHandleExpireOperation, AutoHandleExpireOperationInfo, ActivityExecutorParticipantExpression, SmsMessageTemplate, SmsMessageVariable, } from "@gsp-wf/wf-process-model";
import { mxEventObject, mxResources } from "@edp-pmf/mxgraph-ts";
import { MessageReceiver } from "../entity/MessageReceiver";
import { BizComponent } from "../entity/BizComponent";
import { BizParameter } from "../entity/BizParameter";
import { WfDesignerUtils } from "../WfDesignerUtils";
import { WfEvents } from "../WfEvents";
import { ProcessPropertyKeys } from "./property-keys/ProcessPropertyKeys";
import { PropertyHelper } from "./PropertyHelper";
/**
 * 流程元素类。
 * 取名虽然Cell，但实际上不需要继承mxCell以及Cell类，仅持有流程与画布实例，方法属性获取与修改操作。
 */
var /**
 * 流程元素类。
 * 取名虽然Cell，但实际上不需要继承mxCell以及Cell类，仅持有流程与画布实例，方法属性获取与修改操作。
 */
WfProcessCell = /** @class */ (function (_super) {
    tslib_1.__extends(WfProcessCell, _super);
    function WfProcessCell() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    WfProcessCell.prototype.getFarrisPropConfig = /**
     * @return {?}
     */
    function () {
        var e_1, _a, e_2, _b, e_3, _c;
        /** @type {?} */
        var extendList = [];
        if (this.editorUi.procExtProps && this.editorUi.procExtProps.length > 0) {
            extendList = JSON.parse(JSON.stringify(this.editorUi.procExtProps)).sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) {
                return a.sortOrder - b.sortOrder;
            }));
        }
        extendList = extendList.sort((/**
         * @param {?} a
         * @param {?} b
         * @return {?}
         */
        function (a, b) {
            return a.sortOrder - b.sortOrder;
        }));
        /** @type {?} */
        var propertyData = {};
        /** @type {?} */
        var usual = new ElementPropertyConfig();
        usual.categoryId = 'usual';
        usual.categoryName = mxResources.get("basicInfo");
        usual.tabId = 'basic';
        usual.tabName = mxResources.get("basic");
        usual.properties = [
            { propertyID: ProcessPropertyKeys.Name, propertyName: mxResources.get("name"), propertyType: 'string', visible: !this.editorUi.multiLangEnabled },
            { propertyID: ProcessPropertyKeys.NameLanguage, propertyName: mxResources.get("nameLanguage"), propertyType: 'multiLanguage', visible: this.editorUi.multiLangEnabled },
            { propertyID: ProcessPropertyKeys.Key, propertyName: mxResources.get("key"), propertyType: 'string', readonly: true },
            { propertyID: ProcessPropertyKeys.EntryForm, propertyName: mxResources.get("entryForm"), propertyType: 'modal' },
            { propertyID: ProcessPropertyKeys.startMode, propertyName: mxResources.get("startMode"), propertyType: 'select' },
            { propertyID: ProcessPropertyKeys.wf_TitleExpression, propertyName: mxResources.get("wfTitle"), propertyType: 'modal' },
        ];
        /** @type {?} */
        var formDef = new ElementPropertyConfig();
        formDef.categoryId = 'formDef';
        formDef.categoryName = mxResources.get("formDef");
        formDef.tabId = 'basic';
        formDef.tabName = mxResources.get("basic");
        formDef.properties = [
            { propertyID: ProcessPropertyKeys.wf_ExecuteForms_PC, propertyName: mxResources.get("formsPC"), propertyType: 'modal' },
            { propertyID: ProcessPropertyKeys.wf_ExecuteForms_Mobile, propertyName: mxResources.get("formsMobile"), propertyType: 'modal' },
        ];
        /** @type {?} */
        var process = new ElementPropertyConfig();
        process.categoryId = 'process';
        process.categoryName = mxResources.get("process");
        process.tabId = 'advanced';
        process.tabName = mxResources.get("advanced");
        process.properties = [
            { propertyID: ProcessPropertyKeys.wf_Component, propertyName: mxResources.get("wfComponent"), propertyType: 'modal' },
        ];
        /** @type {?} */
        var advanced = new ElementPropertyConfig();
        advanced.categoryId = 'advanced';
        advanced.categoryName = mxResources.get("otherConfig");
        advanced.tabId = 'advanced';
        advanced.tabName = mxResources.get("advanced");
        advanced.properties = [
            { propertyID: ProcessPropertyKeys.processVariables, propertyName: mxResources.get("processVariables"), propertyType: 'modal' },
            { propertyID: ProcessPropertyKeys.wf_CustomizeAction, propertyName: mxResources.get("customizeAction"), propertyType: 'modal' },
            { propertyID: ProcessPropertyKeys.enablePriority, propertyName: mxResources.get("enablePriority"), propertyType: 'select' },
            { propertyID: ProcessPropertyKeys.enableWholeOrderRetrieve, propertyName: mxResources.get("enableWholeOrderRetrieve"), propertyType: 'select', description: mxResources.get("description") },
            { propertyID: ProcessPropertyKeys.imageSettings, propertyName: mxResources.get("messageProperty"), propertyType: 'modal' },
        ];
        /** @type {?} */
        var time = new ElementPropertyConfig();
        time.categoryId = 'messageProperty'; // 消息属性
        time.categoryName = mxResources.get("timelinessInfo");
        time.tabId = 'advanced';
        time.tabName = mxResources.get("advanced");
        time.properties = [
            { propertyID: ProcessPropertyKeys.timeLimitControl, propertyName: mxResources.get("timeLimitControl"), propertyType: 'modal' },
            { propertyID: ProcessPropertyKeys.messageNotice, propertyName: mxResources.get("messageNotice"), propertyType: 'modal' },
        ];
        // 扩展属性分group处理，每个group包含父级属性和属于它的子扩展属性
        /** @type {?} */
        var groupList = new Array();
        try {
            for (var extendList_1 = tslib_1.__values(extendList), extendList_1_1 = extendList_1.next(); !extendList_1_1.done; extendList_1_1 = extendList_1.next()) {
                var extend = extendList_1_1.value;
                if (!extend.parentId) {
                    /** @type {?} */
                    var group = { parent: extend, childs: new Array() };
                    try {
                        for (var extendList_2 = tslib_1.__values(extendList), extendList_2_1 = extendList_2.next(); !extendList_2_1.done; extendList_2_1 = extendList_2.next()) {
                            var e = extendList_2_1.value;
                            if (e.parentId === extend.id) {
                                group.childs.push(e);
                            }
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (extendList_2_1 && !extendList_2_1.done && (_b = extendList_2.return)) _b.call(extendList_2);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                    groupList.push(group);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (extendList_1_1 && !extendList_1_1.done && (_a = extendList_1.return)) _a.call(extendList_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        // 没有扩展属性值的时候不显示扩展属性一栏
        if (groupList && groupList.length > 0) {
            /** @type {?} */
            var extendProperty = new ElementPropertyConfig();
            extendProperty.categoryId = 'extendProperty';
            extendProperty.categoryName = mxResources.get("extendProperty");
            extendProperty.tabId = 'advanced';
            extendProperty.tabName = mxResources.get("advanced");
            extendProperty.properties = new Array();
            try {
                // 组织扩展属性的二级list
                for (var groupList_1 = tslib_1.__values(groupList), groupList_1_1 = groupList_1.next(); !groupList_1_1.done; groupList_1_1 = groupList_1.next()) {
                    var g = groupList_1_1.value;
                    extendProperty.properties.push({ propertyID: g.parent.id, propertyName: g.parent.propertyTitle, propertyType: 'modal' });
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (groupList_1_1 && !groupList_1_1.done && (_c = groupList_1.return)) _c.call(groupList_1);
                }
                finally { if (e_3) throw e_3.error; }
            }
            this.setProcessPropertyValue(propertyData, groupList);
            return {
                propertyConfig: [usual, formDef, process, time, advanced, extendProperty],
                propertyData: propertyData
            };
        }
        else {
            this.setProcessPropertyValue(propertyData, groupList);
            return {
                propertyConfig: [usual, formDef, process, time, advanced],
                propertyData: propertyData
            };
        }
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    WfProcessCell.prototype.updateProps = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var propertyID = obj.propertyID;
        if (propertyID) {
            switch (propertyID) {
                case 'EntryForm': //流程分类
                    // 更新流程分类名称（流程分类名称可能修改）
                    this.updateProcessHeader(obj.propertyValue.name, obj.propertyValue.businessObjectId, obj.propertyValue.id);
                    this.updateFlowFormInfo(obj.formalParameters, obj.processBizComponnets, obj.executeForms, obj.customizeActions, obj.bizCmps);
                    this.bindProcessTitleExpression(obj.flowForm.processDescription);
                    break;
                case ProcessPropertyKeys.wf_Component: // 流程构件
                    this.updateProcessComponents(obj.propertyValue);
                    break;
                case ProcessPropertyKeys.wf_ExecuteForms_PC:
                    this.updateExecuteForm(obj.propertyValue, 'PC');
                    if (obj.parameters && obj.parameters.length > 0) {
                        PropertyHelper.updateFormalParameters(this.process, obj.parameters);
                    }
                    break;
                case ProcessPropertyKeys.wf_ExecuteForms_Mobile:
                    this.updateExecuteForm(obj.propertyValue, 'MOBILE');
                    if (obj.parameters && obj.parameters.length > 0) {
                        PropertyHelper.updateFormalParameters(this.process, obj.parameters);
                    }
                    break;
                case ProcessPropertyKeys.wf_CustomizeAction:
                    this.updateCustomizeActions(obj.propertyValue);
                    break;
                case ProcessPropertyKeys.wf_TitleExpression:
                    this.modifyProcessTitleExpression(obj.propertyValue);
                    break;
                case ProcessPropertyKeys.allowPreAssign:
                    this.updateAllowPreAssign(obj.propertyValue);
                    break;
                case ProcessPropertyKeys.startMode:
                    this.updateStartMode(obj.propertyValue, obj.allBizCmps, obj.bizCmps);
                    break;
                case ProcessPropertyKeys.enablePriority:
                    this.updateEnablePriority(obj.propertyValue);
                    break;
                case ProcessPropertyKeys.enableWholeOrderRetrieve:
                    this.updateEnableWholeOrderRetrieve(obj.propertyValue);
                    break;
                case ProcessPropertyKeys.processVariables:
                    this.updateProcessVariables(obj.propertyValue);
                    break;
                case ProcessPropertyKeys.timeLimitControl:
                    this.updateProcessTimeLimitControl(obj.propertyValue);
                    break;
                case ProcessPropertyKeys.messageNotice:
                    this.updateProcessMessageNotices(obj.propertyValue);
                    break;
                case ProcessPropertyKeys.imageSettings:
                    this.updateImageSettings(obj.propertyValue);
                    break;
                default:
                    // 判断ResourceKeys是不是groupid,是就走resetExtendProperty
                    if (this.editorUi.procExtProps.findIndex((/**
                     * @param {?} p
                     * @return {?}
                     */
                    function (p) { return p.id === propertyID; })) > -1) {
                        this.resetExtendProperty(obj.propertyValue);
                    }
                    else {
                        _super.prototype.updateProps.call(this, obj);
                    }
                    break;
            }
        }
    };
    /**
    * 设置流程属性值
    * @param bpmnModel
    * @param propertyData
    * @param groupList
    */
    /**
     * 设置流程属性值
     * @param {?} propertyData
     * @param {?=} groupList
     * @return {?}
     */
    WfProcessCell.prototype.setProcessPropertyValue = /**
     * 设置流程属性值
     * @param {?} propertyData
     * @param {?=} groupList
     * @return {?}
     */
    function (propertyData, groupList) {
        var e_4, _a;
        /** @type {?} */
        var process = this.process;
        if (process) {
            if (!process.Name)
                process.Name = '默认流程';
            if (!process.Key)
                process.Key = BpmnModelHelper.GenerateElementId();
            propertyData[ProcessPropertyKeys.Id] = process.Id;
            propertyData[ProcessPropertyKeys.ClrTypeID] = process.ClrTypeID;
            propertyData[ProcessPropertyKeys.Name] = process.Name;
            propertyData[ProcessPropertyKeys.NameLanguage] = process.NameLanguage || {
                "zh-CHS": process.Name,
            };
            propertyData[ProcessPropertyKeys.Key] = process.Key;
            propertyData[ProcessPropertyKeys.wf_TitleExpression] = '';
            /** @type {?} */
            var extendElements = process.GetExtensionElements();
            /** @type {?} */
            var component = [];
            /** @type {?} */
            var webForm = void 0;
            /** @type {?} */
            var mobileForm = void 0;
            /** @type {?} */
            var customizeActions = [];
            /** @type {?} */
            var proExtendProperties_1 = [];
            /** @type {?} */
            var imageSettings = void 0;
            /** @type {?} */
            var messagesNotices = new Array();
            // tslint:disable-next-line: forin
            for (var i in extendElements) {
                if (extendElements[i] instanceof ProcessHeader) {
                    propertyData[ProcessPropertyKeys.EntryForm] = {
                        id: ((/** @type {?} */ (extendElements[i]))).BizDefKey,
                        name: ((/** @type {?} */ (extendElements[i]))).BizDefName
                    };
                    propertyData[ProcessPropertyKeys.allowPreAssign] = ((/** @type {?} */ (extendElements[i]))).allowPreAssign;
                    propertyData[ProcessPropertyKeys.startMode] = ((/** @type {?} */ (extendElements[i]))).startMode;
                    propertyData[ProcessPropertyKeys.enablePriority] = ((/** @type {?} */ (extendElements[i]))).enablePriority ? ((/** @type {?} */ (extendElements[i]))).enablePriority : false;
                    propertyData[ProcessPropertyKeys.enableWholeOrderRetrieve] = ((/** @type {?} */ (extendElements[i]))).enableWholeOrderRetrieve ? ((/** @type {?} */ (extendElements[i]))).enableWholeOrderRetrieve : false;
                }
                if (extendElements[i] instanceof Component
                    && ((/** @type {?} */ (extendElements[i]))).Time !== ToolTime.UNDEFINITION) {
                    component.push(this.convert2Component((/** @type {?} */ (extendElements[i]))));
                }
                if (extendElements[i] instanceof ExecuteForm) {
                    /** @type {?} */
                    var form = (/** @type {?} */ (extendElements[i]));
                    if (form.TerminalType === 'PC') {
                        webForm = form;
                    }
                    else if (form.TerminalType === 'MOBILE') {
                        mobileForm = form;
                    }
                }
                if (extendElements[i] instanceof TitleExpression) {
                    propertyData[ProcessPropertyKeys.wf_TitleExpression] =
                        ((/** @type {?} */ (extendElements[i]))).valueLanguage ? ((/** @type {?} */ (extendElements[i]))).valueLanguage : ((/** @type {?} */ (extendElements[i]))).value;
                }
                if (extendElements[i] instanceof CustomizeAction) {
                    customizeActions.push(extendElements[i]);
                }
                if (extendElements[i] instanceof MessageNotice) {
                    messagesNotices.push(this.getMessageNotice((/** @type {?} */ (extendElements[i]))));
                }
                if (extendElements[i] instanceof TimeLimitControl) {
                    propertyData[ProcessPropertyKeys.timeLimitControl] = this.getPrcocessTimeLimitInfo((/** @type {?} */ (extendElements[i])));
                }
                if (extendElements[i] instanceof ExtendProperty) {
                    proExtendProperties_1.push(extendElements[i]);
                }
                if (extendElements[i] instanceof ImageSettings) {
                    // imageSettings.push(extendElements[i]);
                    propertyData[ProcessPropertyKeys.imageSettings] = extendElements[i];
                }
            }
            /** @type {?} */
            var processVariables = this.getProcessVariables(process);
            propertyData[ProcessPropertyKeys.processVariables] = processVariables;
            propertyData[ProcessPropertyKeys.wf_Component] = component;
            propertyData[ProcessPropertyKeys.wf_ExecuteForms_PC] = webForm;
            propertyData[ProcessPropertyKeys.wf_ExecuteForms_Mobile] = mobileForm;
            propertyData[ProcessPropertyKeys.wf_CustomizeAction] = customizeActions;
            propertyData[ProcessPropertyKeys.messageNotice] = messagesNotices;
            try {
                for (var groupList_2 = tslib_1.__values(groupList), groupList_2_1 = groupList_2.next(); !groupList_2_1.done; groupList_2_1 = groupList_2.next()) {
                    var g = groupList_2_1.value;
                    // 取回模型上的value值给到扩展属性组件里
                    if (proExtendProperties_1 && proExtendProperties_1.length > 0) {
                        g.childs.sort((/**
                         * @param {?} a
                         * @param {?} b
                         * @return {?}
                         */
                        function (a, b) {
                            return a.sortOrder - b.sortOrder;
                        })).forEach((/**
                         * @param {?} d
                         * @return {?}
                         */
                        function (d) {
                            proExtendProperties_1.forEach((/**
                             * @param {?} e
                             * @return {?}
                             */
                            function (e) {
                                if (e.key == d.code) {
                                    d.value = e.value;
                                }
                            }));
                        }));
                    }
                    propertyData[g.parent.id] = g.childs;
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (groupList_2_1 && !groupList_2_1.done && (_a = groupList_2.return)) _a.call(groupList_2);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
    };
    /**
     * @param {?} process
     * @return {?}
     */
    WfProcessCell.prototype.getProcessVariables = /**
     * @param {?} process
     * @return {?}
     */
    function (process) {
        var e_5, _a, e_6, _b, e_7, _c;
        /** @type {?} */
        var processVariables = new Array();
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        /** @type {?} */
        var dataObjects = process.GetDataObjects();
        if (dataObjects !== null && dataObjects.length > 0) {
            try {
                for (var dataObjects_1 = tslib_1.__values(dataObjects), dataObjects_1_1 = dataObjects_1.next(); !dataObjects_1_1.done; dataObjects_1_1 = dataObjects_1.next()) {
                    var obj = dataObjects_1_1.value;
                    if (obj instanceof ValuedDataObject && obj.DataType.Category !== DataTypes.CustomType) {
                        /** @type {?} */
                        var variable = {
                            id: obj.Id,
                            name: obj.Name,
                            documentation: obj.Documentation,
                            basicType: obj.DataType.BasicType,
                            tag: obj.variableSetting ? obj.variableSetting.tag : '',
                            value: obj.variableSetting ? obj.variableSetting.value : '',
                            valueText: obj.variableSetting ? obj.variableSetting.valueText : '',
                            description: obj.DataType.Description
                        };
                        if (obj.variableSetting && obj.variableSetting.tag === 'component') {
                            if (extendElements.length > 0) {
                                try {
                                    for (var extendElements_1 = tslib_1.__values(extendElements), extendElements_1_1 = extendElements_1.next(); !extendElements_1_1.done; extendElements_1_1 = extendElements_1.next()) {
                                        var e = extendElements_1_1.value;
                                        if (e instanceof Component && e.Id === obj.variableSetting.value) {
                                            /** @type {?} */
                                            var component = {
                                                Id: e.Id,
                                                Name: e.Name,
                                                MetadataId: e.MetadataId,
                                                MethodCode: e.MethodCode,
                                                ActualParameters: []
                                            };
                                            if (e.ActualParameters && e.ActualParameters.length > 0) {
                                                try {
                                                    for (var _d = tslib_1.__values(e.ActualParameters), _e = _d.next(); !_e.done; _e = _d.next()) {
                                                        var p = _e.value;
                                                        /** @type {?} */
                                                        var parameter = {
                                                            Name: p.Name,
                                                            Value: p.Value,
                                                            Tag: p.Tag,
                                                        };
                                                        component.ActualParameters.push(parameter);
                                                    }
                                                }
                                                catch (e_7_1) { e_7 = { error: e_7_1 }; }
                                                finally {
                                                    try {
                                                        if (_e && !_e.done && (_c = _d.return)) _c.call(_d);
                                                    }
                                                    finally { if (e_7) throw e_7.error; }
                                                }
                                            }
                                            variable.component = component;
                                        }
                                    }
                                }
                                catch (e_6_1) { e_6 = { error: e_6_1 }; }
                                finally {
                                    try {
                                        if (extendElements_1_1 && !extendElements_1_1.done && (_b = extendElements_1.return)) _b.call(extendElements_1);
                                    }
                                    finally { if (e_6) throw e_6.error; }
                                }
                            }
                        }
                        processVariables.push(variable);
                    }
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (dataObjects_1_1 && !dataObjects_1_1.done && (_a = dataObjects_1.return)) _a.call(dataObjects_1);
                }
                finally { if (e_5) throw e_5.error; }
            }
        }
        return processVariables;
    };
    /**
     * @param {?} timeLimitControl
     * @return {?}
     */
    WfProcessCell.prototype.getPrcocessTimeLimitInfo = /**
     * @param {?} timeLimitControl
     * @return {?}
     */
    function (timeLimitControl) {
        var e_8, _a, e_9, _b;
        /** @type {?} */
        var timeLimit;
        if (timeLimitControl) {
            timeLimit = new TimeLimitControlInfo();
            timeLimit.enable = timeLimitControl.enable;
            timeLimit.tag = timeLimitControl.tag;
            timeLimit.dueTime = timeLimitControl.dueTime;
            timeLimit.timeUnit = timeLimitControl.timeUnit;
            timeLimit.expireOperations = new Array();
            if (timeLimitControl.expireOperations && timeLimitControl.expireOperations.length > 0) {
                /** @type {?} */
                var operations = timeLimitControl.expireOperations;
                try {
                    for (var operations_1 = tslib_1.__values(operations), operations_1_1 = operations_1.next(); !operations_1_1.done; operations_1_1 = operations_1.next()) {
                        var o = operations_1_1.value;
                        if (o.ClrTypeID.indexOf('SendMessageExpireOperation') > 0) {
                            /** @type {?} */
                            var operation = (/** @type {?} */ (o));
                            /** @type {?} */
                            var sendMessage = new SendMessageExpireOperationInfo();
                            sendMessage.timeDuration = operation.timeDuration;
                            sendMessage.durationType = operation.durationType;
                            sendMessage.timeUnit = operation.timeUnit;
                            sendMessage.message = operation.message;
                            sendMessage.sendType = operation.sendType;
                            sendMessage.subject = operation.subject;
                            sendMessage.id = operation.id;
                            sendMessage.receiver = operation.receiver;
                            sendMessage.receiverText = operation.receiverText;
                            sendMessage.cycleEnable = operation.cycleEnable;
                            sendMessage.cycleTime = operation.cycleTime;
                            sendMessage.cycleTimeUnit = operation.cycleTimeUnit;
                            sendMessage.operationTypeName = operation.operationTypeName ? operation.operationTypeName : 'SendMessageExpireOperation';
                            if (operation.smsMessageTemplate) {
                                /** @type {?} */
                                var smsTemplate = new Object;
                                smsTemplate.code = operation.smsMessageTemplate.code;
                                smsTemplate.content = operation.smsMessageTemplate.content;
                                smsTemplate.smsMessageVariables = new Array();
                                if (operation.smsMessageTemplate.smsMessageVariables && operation.smsMessageTemplate.smsMessageVariables.length > 0) {
                                    try {
                                        for (var _c = tslib_1.__values(operation.smsMessageTemplate.smsMessageVariables), _d = _c.next(); !_d.done; _d = _c.next()) {
                                            var v = _d.value;
                                            /** @type {?} */
                                            var variable = new Object;
                                            variable.id = WfDesignerUtils.createNM();
                                            variable.code = v.code;
                                            variable.value = v.value;
                                            variable.tag = v.tag;
                                            smsTemplate.smsMessageVariables.push(variable);
                                        }
                                    }
                                    catch (e_9_1) { e_9 = { error: e_9_1 }; }
                                    finally {
                                        try {
                                            if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                                        }
                                        finally { if (e_9) throw e_9.error; }
                                    }
                                }
                                sendMessage.smsMessageTemplate = smsTemplate;
                            }
                            timeLimit.expireOperations.push(sendMessage);
                        }
                        else if (o.ClrTypeID.indexOf('AutoHandleExpireOperation') > 0) {
                            /** @type {?} */
                            var operation = (/** @type {?} */ (o));
                            /** @type {?} */
                            var autoHandle = new AutoHandleExpireOperationInfo;
                            autoHandle.id = operation.id;
                            autoHandle.durationType = operation.durationType;
                            autoHandle.timeDuration = operation.timeDuration;
                            autoHandle.timeUnit = operation.timeUnit;
                            autoHandle.operationTypeName = operation.operationTypeName ? operation.operationTypeName : 'AutoHandleExpireOperation';
                            autoHandle.autoHandleType = operation.autoHandleType;
                            autoHandle.message = operation.message;
                            timeLimit.expireOperations.push(autoHandle);
                        }
                    }
                }
                catch (e_8_1) { e_8 = { error: e_8_1 }; }
                finally {
                    try {
                        if (operations_1_1 && !operations_1_1.done && (_a = operations_1.return)) _a.call(operations_1);
                    }
                    finally { if (e_8) throw e_8.error; }
                }
            }
        }
        return timeLimit;
    };
    /**
     * @param {?} messageNotice
     * @return {?}
     */
    WfProcessCell.prototype.getMessageNotice = /**
     * @param {?} messageNotice
     * @return {?}
     */
    function (messageNotice) {
        var e_10, _a, e_11, _b;
        /** @type {?} */
        var message = new Object;
        message.content = messageNotice.content;
        message.sendTime = messageNotice.sendTime;
        message.sendType = messageNotice.sendType;
        message.subject = messageNotice.subject;
        if (messageNotice.smsMessageTemplate) {
            /** @type {?} */
            var smsTemplate = new Object;
            smsTemplate.code = messageNotice.smsMessageTemplate.code;
            smsTemplate.content = messageNotice.smsMessageTemplate.content;
            smsTemplate.smsMessageVariables = new Array();
            if (messageNotice.smsMessageTemplate.smsMessageVariables && messageNotice.smsMessageTemplate.smsMessageVariables.length > 0) {
                try {
                    for (var _c = tslib_1.__values(messageNotice.smsMessageTemplate.smsMessageVariables), _d = _c.next(); !_d.done; _d = _c.next()) {
                        var v = _d.value;
                        /** @type {?} */
                        var variable = new Object;
                        variable.id = WfDesignerUtils.createNM();
                        variable.code = v.code;
                        variable.value = v.value;
                        variable.tag = v.tag;
                        smsTemplate.smsMessageVariables.push(variable);
                    }
                }
                catch (e_10_1) { e_10 = { error: e_10_1 }; }
                finally {
                    try {
                        if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                    }
                    finally { if (e_10) throw e_10.error; }
                }
            }
            message.smsMessageTemplate = smsTemplate;
        }
        /** @type {?} */
        var receiverExpressions = new Array();
        if (messageNotice.receiver) {
            /** @type {?} */
            var receivers = messageNotice.receiver.split(',');
            /** @type {?} */
            var receiverTexts = messageNotice.receiverText.split(',');
            /** @type {?} */
            var userReceiver = new MessageReceiver();
            userReceiver.type = 'userParticipant';
            userReceiver.userIds = '';
            userReceiver.userNames = '';
            for (var i in receivers) {
                if (receivers[i] === 'procInstCreator') {
                    /** @type {?} */
                    var procInstCreator = new MessageReceiver();
                    procInstCreator.type = 'processParticipant';
                    procInstCreator.processParticipant = 'procInstCreator';
                    receiverExpressions.push(procInstCreator);
                }
                else if (receivers[i] === 'workItemParticipant') {
                    /** @type {?} */
                    var workItemParticipant = new MessageReceiver();
                    workItemParticipant.type = 'processParticipant';
                    workItemParticipant.processParticipant = 'workItemParticipant';
                    receiverExpressions.push(workItemParticipant);
                }
                else {
                    userReceiver.userIds += (receivers[i] + ',');
                    userReceiver.userNames += (receiverTexts[i] + ',');
                }
            }
            if (userReceiver.userIds) {
                userReceiver.userIds = userReceiver.userIds.substring(0, userReceiver.userIds.length - 1);
                userReceiver.userNames = userReceiver.userNames.substring(0, userReceiver.userNames.length - 1);
                receiverExpressions.push(userReceiver);
            }
        }
        if (messageNotice.receiverExpressions && messageNotice.receiverExpressions.length > 0) {
            try {
                for (var _e = tslib_1.__values(messageNotice.receiverExpressions), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var e = _f.value;
                    if (e instanceof ActivityExecutorParticipantExpression) {
                        /** @type {?} */
                        var activityExecutorParticipant = new MessageReceiver();
                        activityExecutorParticipant.type = 'processParticipant';
                        activityExecutorParticipant.processParticipant = 'activityExecutorParticipant';
                        activityExecutorParticipant.activityDefId = ((/** @type {?} */ (e))).activityDefId;
                        activityExecutorParticipant.activityDefName = ((/** @type {?} */ (e))).activityDefName;
                        receiverExpressions.push(activityExecutorParticipant);
                    }
                    else if (e instanceof UserParticipantExpression) {
                        /** @type {?} */
                        var variableReceiver = new MessageReceiver();
                        variableReceiver.type = 'variableParticipant';
                        variableReceiver.variableParticipant = ((/** @type {?} */ (e))).UserId;
                        variableReceiver.variableName = ((/** @type {?} */ (e))).name;
                        variableReceiver.orgType = ((/** @type {?} */ (e))).OrgType;
                        receiverExpressions.push(variableReceiver);
                    }
                    else if (e instanceof PositionParticipantExpression) {
                        /** @type {?} */
                        var positionReceiver = new MessageReceiver();
                        positionReceiver.type = 'positionParticipant';
                        positionReceiver.positionId = ((/** @type {?} */ (e))).PositionId;
                        positionReceiver.positionName = ((/** @type {?} */ (e))).PositionName;
                        positionReceiver.positionType = ((/** @type {?} */ (e))).PositionType;
                        if (positionReceiver.positionType === '1') {
                            positionReceiver.orgId = ((/** @type {?} */ (e))).OrgId;
                            positionReceiver.orgType = ((/** @type {?} */ (e))).OrgType;
                        }
                        receiverExpressions.push(positionReceiver);
                    }
                }
            }
            catch (e_11_1) { e_11 = { error: e_11_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                }
                finally { if (e_11) throw e_11.error; }
            }
        }
        if (receiverExpressions.length > 0) {
            message.receivers = receiverExpressions;
        }
        return message;
    };
    /**
     * @param {?} component
     * @return {?}
     */
    WfProcessCell.prototype.convert2Component = /**
     * @param {?} component
     * @return {?}
     */
    function (component) {
        var e_12, _a;
        /** @type {?} */
        var bizComponent = new BizComponent();
        bizComponent.id = component.Id;
        bizComponent.name = component.Name;
        bizComponent.metadataId = component.MetadataId;
        bizComponent.methodCode = component.MethodCode;
        bizComponent.executionTime = component.Time;
        bizComponent.isForCompensation = component.isForCompensation;
        bizComponent.scopeType = component.ScopeType || null;
        bizComponent.actualParameters = new Array();
        if (component.ActualParameters && component.ActualParameters.length > 0) {
            try {
                for (var _b = tslib_1.__values(component.ActualParameters), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var param = _c.value;
                    /** @type {?} */
                    var bizParam = new BizParameter();
                    bizParam.code = param.Name;
                    bizParam.name = param.Name;
                    bizParam.value = param.Value;
                    bizComponent.actualParameters.push(bizParam);
                }
            }
            catch (e_12_1) { e_12 = { error: e_12_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_12) throw e_12.error; }
            }
        }
        return bizComponent;
    };
    /**
     * 设置ProcessHeader
     * @param ffId
     * @param bizObjId
     * @param ffName
     * @param startMode
     */
    /**
     * 设置ProcessHeader
     * @param {?} ffId
     * @param {?} bizObjId
     * @param {?} ffName
     * @param {?} startMode
     * @return {?}
     */
    WfProcessCell.prototype.setProcessHeader = /**
     * 设置ProcessHeader
     * @param {?} ffId
     * @param {?} bizObjId
     * @param {?} ffName
     * @param {?} startMode
     * @return {?}
     */
    function (ffId, bizObjId, ffName, startMode) {
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        /** @type {?} */
        var processHeader = new ProcessHeader(process.Model, ffId, bizObjId, ffName, startMode);
        extendElements.push(processHeader);
    };
    /**
     * 更新流程分类绑定的参数，流程组件，表单信息
     * @param params 参数列表
     * @param cpts 流程组件
     * @param form 流程表单
     */
    /**
     * 更新流程分类绑定的参数，流程组件，表单信息
     * @param {?} params 参数列表
     * @param {?} cpts 流程组件
     * @param {?} forms
     * @param {?} customizeActions
     * @param {?} bizCmps
     * @return {?}
     */
    WfProcessCell.prototype.updateFlowFormInfo = /**
     * 更新流程分类绑定的参数，流程组件，表单信息
     * @param {?} params 参数列表
     * @param {?} cpts 流程组件
     * @param {?} forms
     * @param {?} customizeActions
     * @param {?} bizCmps
     * @return {?}
     */
    function (params, cpts, forms, customizeActions, bizCmps) {
        this.updateFormalParameters(params, cpts.filter((/**
         * @param {?} cpt
         * @return {?}
         */
        function (cpt) { return cpt.ownerType === 'Schema'; })));
        this.setProcessComponents(cpts, bizCmps);
        this.updateExecuteForms(forms);
        this.updateCustomizeActions(customizeActions);
    };
    /**
     * @param {?} parameters
     * @param {?=} cmps
     * @return {?}
     */
    WfProcessCell.prototype.updateFormalParameters = /**
     * @param {?} parameters
     * @param {?=} cmps
     * @return {?}
     */
    function (parameters, cmps) {
        if (cmps === void 0) { cmps = new Array(); }
        var e_13, _a;
        if (!parameters) {
            parameters = [];
        }
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var objects = process.GetDataObjects();
        if (parameters !== null && parameters.length > 0) {
            var _loop_1 = function (p) {
                var e_14, _a, e_15, _b;
                /** @type {?} */
                var dataType = new DataType(process.Model);
                if (p.dataType === 'CDTSchema') {
                    try {
                        // 先清理CDTSchema类型的dataObject
                        for (var objects_1 = tslib_1.__values(objects), objects_1_1 = objects_1.next(); !objects_1_1.done; objects_1_1 = objects_1.next()) {
                            var o = objects_1_1.value;
                            //TODO:数据类型为customType暂时默认为参数预制
                            if (o instanceof DataObject && o.DataType.Category === DataTypes.CustomType) {
                                process.RemoveDataObject(o.Id);
                            }
                        }
                    }
                    catch (e_14_1) { e_14 = { error: e_14_1 }; }
                    finally {
                        try {
                            if (objects_1_1 && !objects_1_1.done && (_a = objects_1.return)) _a.call(objects_1);
                        }
                        finally { if (e_14) throw e_14.error; }
                    }
                    // 重新绑定流程分类支持精简be
                    /** @type {?} */
                    var customType = new CDTSchemaType(process.Model);
                    if (p.newSchemaId && p.newSchemaMetadataId) {
                        customType.SchemaId = p.newSchemaId;
                        customType.MetadataId = p.newSchemaMetadataId;
                    }
                    else {
                        customType.SchemaId = p.schemaId;
                        customType.MetadataId = p.schemaMetadataId;
                    }
                    customType.ComponentId = cmps.filter((/**
                     * @param {?} cmp
                     * @return {?}
                     */
                    function (cmp) { return cmp.owner === p.schemaId; }))[0].id;
                    dataType.Category = DataTypes.CustomType;
                    dataType.CustomType = customType;
                }
                else {
                    try {
                        // 先清理code相同的dataObject
                        for (var objects_2 = tslib_1.__values(objects), objects_2_1 = objects_2.next(); !objects_2_1.done; objects_2_1 = objects_2.next()) {
                            var o = objects_2_1.value;
                            if (o instanceof DataObject && o.Name === p.code) {
                                process.RemoveDataObject(o.Id);
                            }
                        }
                    }
                    catch (e_15_1) { e_15 = { error: e_15_1 }; }
                    finally {
                        try {
                            if (objects_2_1 && !objects_2_1.done && (_b = objects_2.return)) _b.call(objects_2);
                        }
                        finally { if (e_15) throw e_15.error; }
                    }
                    dataType.Category = DataTypes.BasicType;
                    dataType.BasicType = p.dataType;
                }
                dataType.Description = p.description;
                // dataType.InitialValue = p.initialValue;
                if (p.parameterMode === 0) {
                    /** @type {?} */
                    var dataObject = new ValuedDataObject(process.Model);
                    dataObject.Name = p.code;
                    dataObject.Documentation = p.name;
                    dataObject.DataType = dataType;
                    dataObject.IsAllowNull = p.allowNull;
                    if (p.initialValue || p.initialValueType) {
                        /** @type {?} */
                        var variableSetting = new VariableSetting(process.Model);
                        variableSetting.variableCode = p.code;
                        variableSetting.variableName = p.name;
                        variableSetting.tag = p.initialValueType;
                        variableSetting.value = p.initialValue;
                        variableSetting.valueText = p.initialValue;
                        dataObject.variableSetting = variableSetting;
                    }
                    process.AddDataObject(dataObject);
                }
            };
            try {
                for (var parameters_1 = tslib_1.__values(parameters), parameters_1_1 = parameters_1.next(); !parameters_1_1.done; parameters_1_1 = parameters_1.next()) {
                    var p = parameters_1_1.value;
                    _loop_1(p);
                }
            }
            catch (e_13_1) { e_13 = { error: e_13_1 }; }
            finally {
                try {
                    if (parameters_1_1 && !parameters_1_1.done && (_a = parameters_1.return)) _a.call(parameters_1);
                }
                finally { if (e_13) throw e_13.error; }
            }
        }
    };
    /**
     * 基于流程分类的切换更新流程构件
     */
    /**
     * 基于流程分类的切换更新流程构件
     * @param {?} procCpts
     * @param {?} bizCmps
     * @return {?}
     */
    WfProcessCell.prototype.setProcessComponents = /**
     * 基于流程分类的切换更新流程构件
     * @param {?} procCpts
     * @param {?} bizCmps
     * @return {?}
     */
    function (procCpts, bizCmps) {
        var e_16, _a, e_17, _b, e_18, _c, e_19, _d, e_20, _e, e_21, _f, e_22, _g, e_23, _h, e_24, _j, e_25, _k;
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var processHeader;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        //删除流程构件
        if (extendElements.length > 0) {
            for (var i = extendElements.length; i > 0; i--) {
                if (extendElements[i - 1] instanceof Component
                    && ((/** @type {?} */ (extendElements[i - 1]))).Time !== ToolTime.UNDEFINITION) {
                    extendElements.splice(i - 1, 1);
                }
                if (extendElements[i - 1] instanceof ProcessHeader) {
                    processHeader = (/** @type {?} */ (extendElements[i - 1]));
                }
            }
        }
        // 子流程和外部流程的情况先需要绑定取数构件
        if (processHeader.startMode !== StartMode.DIRECT) {
            /** @type {?} */
            var loadDataComponent = procCpts.find((/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return p.executionTime === ToolTime.LOADDATA; }));
            if (loadDataComponent) {
                /** @type {?} */
                var ldCpt = new Component(process.Model);
                ldCpt.Time = (/** @type {?} */ (loadDataComponent.executionTime));
                ldCpt.MetadataId = loadDataComponent.metadataId;
                ldCpt.MethodCode = loadDataComponent.methodCode;
                ldCpt.Id = loadDataComponent.id;
                ldCpt.Name = loadDataComponent.name;
                ldCpt.isForCompensation = loadDataComponent.isForCompensation;
                ldCpt.ActualParameters = [];
                if (loadDataComponent.scopeType) {
                    ldCpt.ScopeType = loadDataComponent.scopeType;
                }
                else {
                    ldCpt.ScopeType = "Undefined";
                }
                if (loadDataComponent.actualParameters && loadDataComponent.actualParameters.length > 0) {
                    try {
                        for (var _l = tslib_1.__values(loadDataComponent.actualParameters), _m = _l.next(); !_m.done; _m = _l.next()) {
                            var item = _m.value;
                            /** @type {?} */
                            var parameter = new ActualParameter(process.Model, item.code, item.value);
                            ldCpt.ActualParameters.push(parameter);
                        }
                    }
                    catch (e_16_1) { e_16 = { error: e_16_1 }; }
                    finally {
                        try {
                            if (_m && !_m.done && (_a = _l.return)) _a.call(_l);
                        }
                        finally { if (e_16) throw e_16.error; }
                    }
                }
                // 把取数构件放进去
                extendElements.push(ldCpt);
                // 如果是外部流程，校验是否集成统一预制的外部流程构件
                if (processHeader.startMode == StartMode.EXTERNAL_PROCESS) {
                    /** @type {?} */
                    var externalCmps = false;
                    /** @type {?} */
                    var cmps = bizCmps;
                    if (cmps && cmps.length > 0) {
                        try {
                            // 此处标记有外部流程通用构件
                            for (var cmps_1 = tslib_1.__values(cmps), cmps_1_1 = cmps_1.next(); !cmps_1_1.done; cmps_1_1 = cmps_1.next()) {
                                var c = cmps_1_1.value;
                                if (c.scopeType === 'ExternalProcess') {
                                    externalCmps = true;
                                    break;
                                }
                            }
                        }
                        catch (e_17_1) { e_17 = { error: e_17_1 }; }
                        finally {
                            try {
                                if (cmps_1_1 && !cmps_1_1.done && (_b = cmps_1.return)) _b.call(cmps_1);
                            }
                            finally { if (e_17) throw e_17.error; }
                        }
                    }
                    // 如果没有外部流程构件
                    if (!externalCmps) {
                        if (procCpts && procCpts.length > 0) {
                            try {
                                for (var procCpts_1 = tslib_1.__values(procCpts), procCpts_1_1 = procCpts_1.next(); !procCpts_1_1.done; procCpts_1_1 = procCpts_1.next()) {
                                    var procCpt = procCpts_1_1.value;
                                    /** @type {?} */
                                    var cpt = new Component(process.Model);
                                    cpt.Time = (/** @type {?} */ (procCpt.executionTime));
                                    cpt.MetadataId = procCpt.metadataId;
                                    cpt.MethodCode = procCpt.methodCode;
                                    cpt.Id = procCpt.id;
                                    cpt.Name = procCpt.name;
                                    cpt.isForCompensation = procCpt.isForCompensation;
                                    cpt.ActualParameters = [];
                                    if (procCpt.scopeType) {
                                        cpt.ScopeType = procCpt.scopeType;
                                    }
                                    else {
                                        cpt.ScopeType = "Undefined";
                                    }
                                    if (procCpt.actualParameters && procCpt.actualParameters.length > 0) {
                                        try {
                                            for (var _o = tslib_1.__values(procCpt.actualParameters), _p = _o.next(); !_p.done; _p = _o.next()) {
                                                var item = _p.value;
                                                /** @type {?} */
                                                var parameter = new ActualParameter(process.Model, item.code, item.value);
                                                cpt.ActualParameters.push(parameter);
                                            }
                                        }
                                        catch (e_19_1) { e_19 = { error: e_19_1 }; }
                                        finally {
                                            try {
                                                if (_p && !_p.done && (_d = _o.return)) _d.call(_o);
                                            }
                                            finally { if (e_19) throw e_19.error; }
                                        }
                                    }
                                    // 避免重复的放入取数构件,不要取数，子流程和直接流程的构件，要未定义，外部，全量的
                                    if (cpt.Time !== ToolTime.LOADDATA
                                        && cpt.ScopeType !== StartMode.DIRECT
                                        && cpt.ScopeType !== StartMode.SUB_PROCESS) {
                                        extendElements.push(cpt);
                                    }
                                }
                            }
                            catch (e_18_1) { e_18 = { error: e_18_1 }; }
                            finally {
                                try {
                                    if (procCpts_1_1 && !procCpts_1_1.done && (_c = procCpts_1.return)) _c.call(procCpts_1);
                                }
                                finally { if (e_18) throw e_18.error; }
                            }
                        }
                    }
                    else {
                        try {
                            for (var procCpts_2 = tslib_1.__values(procCpts), procCpts_2_1 = procCpts_2.next(); !procCpts_2_1.done; procCpts_2_1 = procCpts_2.next()) {
                                var procCpt = procCpts_2_1.value;
                                /** @type {?} */
                                var cpt = new Component(process.Model);
                                cpt.Time = (/** @type {?} */ (procCpt.executionTime));
                                cpt.MetadataId = procCpt.metadataId;
                                cpt.MethodCode = procCpt.methodCode;
                                cpt.Id = procCpt.id;
                                cpt.Name = procCpt.name;
                                cpt.isForCompensation = procCpt.isForCompensation;
                                cpt.ActualParameters = [];
                                if (procCpt.scopeType) {
                                    cpt.ScopeType = procCpt.scopeType;
                                }
                                else {
                                    cpt.ScopeType = "Undefined";
                                }
                                if (procCpt.actualParameters && procCpt.actualParameters.length > 0) {
                                    try {
                                        for (var _q = tslib_1.__values(procCpt.actualParameters), _r = _q.next(); !_r.done; _r = _q.next()) {
                                            var item = _r.value;
                                            /** @type {?} */
                                            var parameter = new ActualParameter(process.Model, item.code, item.value);
                                            cpt.ActualParameters.push(parameter);
                                        }
                                    }
                                    catch (e_21_1) { e_21 = { error: e_21_1 }; }
                                    finally {
                                        try {
                                            if (_r && !_r.done && (_f = _q.return)) _f.call(_q);
                                        }
                                        finally { if (e_21) throw e_21.error; }
                                    }
                                }
                                // 避免重复的放入取数构件,不要取数，子流程和直接流程,未知的构件，外部，全量的
                                if (cpt.Time !== ToolTime.LOADDATA
                                    && cpt.ScopeType !== StartMode.DIRECT
                                    && cpt.ScopeType !== StartMode.SUB_PROCESS
                                    && cpt.ScopeType !== "Undefined") {
                                    extendElements.push(cpt);
                                }
                            }
                        }
                        catch (e_20_1) { e_20 = { error: e_20_1 }; }
                        finally {
                            try {
                                if (procCpts_2_1 && !procCpts_2_1.done && (_e = procCpts_2.return)) _e.call(procCpts_2);
                            }
                            finally { if (e_20) throw e_20.error; }
                        }
                    }
                    // 如果有外部流程构件，同样只保留取数构件，不需要其他处理(取数构件已经放进去了)
                    // 刷新一下属性框
                    this.editorUi.fireEvent(new mxEventObject(PmfEvents.ShowProperty));
                }
                else if (processHeader.startMode == StartMode.SUB_PROCESS) {
                    if (procCpts && procCpts.length > 0) {
                        try {
                            for (var procCpts_3 = tslib_1.__values(procCpts), procCpts_3_1 = procCpts_3.next(); !procCpts_3_1.done; procCpts_3_1 = procCpts_3.next()) {
                                var procCpt = procCpts_3_1.value;
                                /** @type {?} */
                                var cpt = new Component(process.Model);
                                cpt.Time = (/** @type {?} */ (procCpt.executionTime));
                                cpt.MetadataId = procCpt.metadataId;
                                cpt.MethodCode = procCpt.methodCode;
                                cpt.Id = procCpt.id;
                                cpt.Name = procCpt.name;
                                cpt.isForCompensation = procCpt.isForCompensation;
                                cpt.ActualParameters = [];
                                if (procCpt.scopeType) {
                                    cpt.ScopeType = procCpt.scopeType;
                                }
                                else {
                                    cpt.ScopeType = "Undefined";
                                }
                                if (procCpt.actualParameters && procCpt.actualParameters.length > 0) {
                                    try {
                                        for (var _s = tslib_1.__values(procCpt.actualParameters), _t = _s.next(); !_t.done; _t = _s.next()) {
                                            var item = _t.value;
                                            /** @type {?} */
                                            var parameter = new ActualParameter(process.Model, item.code, item.value);
                                            cpt.ActualParameters.push(parameter);
                                        }
                                    }
                                    catch (e_23_1) { e_23 = { error: e_23_1 }; }
                                    finally {
                                        try {
                                            if (_t && !_t.done && (_h = _s.return)) _h.call(_s);
                                        }
                                        finally { if (e_23) throw e_23.error; }
                                    }
                                }
                                // 避免重复的放入取数构件,不要取数，直接流程和外部流程，未定义的构件，要子流程，全量的
                                if (cpt.Time !== ToolTime.LOADDATA
                                    && cpt.ScopeType !== StartMode.DIRECT
                                    && cpt.ScopeType !== StartMode.EXTERNAL_PROCESS
                                    && cpt.ScopeType !== "Undefined") {
                                    extendElements.push(cpt);
                                }
                            }
                        }
                        catch (e_22_1) { e_22 = { error: e_22_1 }; }
                        finally {
                            try {
                                if (procCpts_3_1 && !procCpts_3_1.done && (_g = procCpts_3.return)) _g.call(procCpts_3);
                            }
                            finally { if (e_22) throw e_22.error; }
                        }
                    }
                }
                // 如果是子流程则不需要其他处理(取数构件已经放进去了)
            }
        }
        else { // 直接流程
            // 直接流程
            // 直接流程过滤掉共享的构件
            /** @type {?} */
            var actualCmps = procCpts.filter((/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return WfDesignerUtils.isPfComponent(p.metadataId) === false; }));
            if (actualCmps && actualCmps.length > 0) {
                try {
                    for (var actualCmps_1 = tslib_1.__values(actualCmps), actualCmps_1_1 = actualCmps_1.next(); !actualCmps_1_1.done; actualCmps_1_1 = actualCmps_1.next()) {
                        var procCpt = actualCmps_1_1.value;
                        /** @type {?} */
                        var cpt = new Component(process.Model);
                        cpt.Time = (/** @type {?} */ (procCpt.executionTime));
                        cpt.MetadataId = procCpt.metadataId;
                        cpt.MethodCode = procCpt.methodCode;
                        cpt.Id = procCpt.id;
                        cpt.Name = procCpt.name;
                        cpt.isForCompensation = procCpt.isForCompensation;
                        if (procCpt.scopeType) {
                            cpt.ScopeType = procCpt.scopeType;
                        }
                        else {
                            cpt.ScopeType = "Undefined";
                        }
                        cpt.ActualParameters = [];
                        if (procCpt.actualParameters && procCpt.actualParameters.length > 0) {
                            try {
                                for (var _u = tslib_1.__values(procCpt.actualParameters), _v = _u.next(); !_v.done; _v = _u.next()) {
                                    var item = _v.value;
                                    /** @type {?} */
                                    var parameter = new ActualParameter(process.Model, item.code, item.value);
                                    cpt.ActualParameters.push(parameter);
                                }
                            }
                            catch (e_25_1) { e_25 = { error: e_25_1 }; }
                            finally {
                                try {
                                    if (_v && !_v.done && (_k = _u.return)) _k.call(_u);
                                }
                                finally { if (e_25) throw e_25.error; }
                            }
                        }
                        // 避免重复的放入取数构件,子流程和外部流程的构件，要未定义，直接，全量的
                        if (cpt.ScopeType !== StartMode.EXTERNAL_PROCESS
                            && cpt.ScopeType !== StartMode.SUB_PROCESS) {
                            extendElements.push(cpt);
                        }
                    }
                }
                catch (e_24_1) { e_24 = { error: e_24_1 }; }
                finally {
                    try {
                        if (actualCmps_1_1 && !actualCmps_1_1.done && (_j = actualCmps_1.return)) _j.call(actualCmps_1);
                    }
                    finally { if (e_24) throw e_24.error; }
                }
            }
        }
    };
    /**
     * @param {?} forms
     * @return {?}
     */
    WfProcessCell.prototype.updateExecuteForms = /**
     * @param {?} forms
     * @return {?}
     */
    function (forms) {
        var e_26, _a;
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        //删除表单构件
        if (extendElements.length > 0) {
            for (var i = extendElements.length; i > 0; i--) {
                if (extendElements[i - 1] instanceof ExecuteForm) {
                    extendElements.splice(i - 1, 1);
                }
            }
        }
        try {
            for (var forms_1 = tslib_1.__values(forms), forms_1_1 = forms_1.next(); !forms_1_1.done; forms_1_1 = forms_1.next()) {
                var f = forms_1_1.value;
                this.updateExecuteForm(f);
            }
        }
        catch (e_26_1) { e_26 = { error: e_26_1 }; }
        finally {
            try {
                if (forms_1_1 && !forms_1_1.done && (_a = forms_1.return)) _a.call(forms_1);
            }
            finally { if (e_26) throw e_26.error; }
        }
    };
    /**
     * @param {?} form
     * @param {?=} terminal
     * @return {?}
     */
    WfProcessCell.prototype.updateExecuteForm = /**
     * @param {?} form
     * @param {?=} terminal
     * @return {?}
     */
    function (form, terminal) {
        var e_27, _a;
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        //删除表单构件
        if (extendElements.length > 0) {
            for (var i = extendElements.length; i > 0; i--) {
                if (extendElements[i - 1] instanceof ExecuteForm) {
                    /** @type {?} */
                    var f = (/** @type {?} */ (extendElements[i - 1]));
                    if (f.TerminalType === (terminal ? terminal : form.TerminalType)) {
                        extendElements.splice(i - 1, 1);
                    }
                }
            }
        }
        if (form) {
            /** @type {?} */
            var params = (/** @type {?} */ (form.ActualParameters));
            /** @type {?} */
            var actParas = new Array();
            if (params && params.length > 0) {
                try {
                    for (var params_1 = tslib_1.__values(params), params_1_1 = params_1.next(); !params_1_1.done; params_1_1 = params_1.next()) {
                        var para = params_1_1.value;
                        /** @type {?} */
                        var actPara = new ActualParameter(process.Model, para.Name, para.Value);
                        if (para.Tag) {
                            actPara.Tag = para.Tag;
                        }
                        else {
                            if (para.Value) {
                                if (para.Value.length >= 8 && para.Value.slice(0, 8) === "{\"expr\":") {
                                    actPara.Tag = 'expression';
                                }
                                else {
                                    actPara.Tag = 'constant';
                                }
                            }
                            else {
                                actPara.Tag = 'variable';
                            }
                        }
                        actParas.push(actPara);
                    }
                }
                catch (e_27_1) { e_27 = { error: e_27_1 }; }
                finally {
                    try {
                        if (params_1_1 && !params_1_1.done && (_a = params_1.return)) _a.call(params_1);
                    }
                    finally { if (e_27) throw e_27.error; }
                }
            }
            /** @type {?} */
            var id = BpmnModelHelper.GenerateElementId();
            /** @type {?} */
            var item = new ExecuteForm(process.Model, id, form.Name, actParas);
            item.ComponentId = form.ComponentId;
            item.FormType = form.FormType;
            item.formFormatId = form.formFormatId;
            item.formatConfigId = form.formatConfigId;
            item.TerminalType = form.TerminalType;
            extendElements.push(item);
        }
    };
    /**
     * @param {?} customizeActions
     * @return {?}
     */
    WfProcessCell.prototype.updateCustomizeActions = /**
     * @param {?} customizeActions
     * @return {?}
     */
    function (customizeActions) {
        var e_28, _a, e_29, _b;
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        //删除流程构件
        if (extendElements.length > 0) {
            for (var i = extendElements.length; i > 0; i--) {
                if (extendElements[i - 1] instanceof CustomizeAction) {
                    extendElements.splice(i - 1, 1);
                }
            }
        }
        //重新绑定
        if (customizeActions && customizeActions.length > 0) {
            try {
                for (var customizeActions_1 = tslib_1.__values(customizeActions), customizeActions_1_1 = customizeActions_1.next(); !customizeActions_1_1.done; customizeActions_1_1 = customizeActions_1.next()) {
                    var a = customizeActions_1_1.value;
                    /** @type {?} */
                    var action = new CustomizeAction(process.Model);
                    action.id = BpmnModelHelper.GenerateElementId();
                    action.code = a.code;
                    action.name = a.name;
                    action.terminalType = a.terminalType;
                    // action.enable = a.enable;
                    action.actualParameters = [];
                    if (a.actualParameters && a.actualParameters.length > 0) {
                        try {
                            for (var _c = tslib_1.__values(a.actualParameters), _d = _c.next(); !_d.done; _d = _c.next()) {
                                var p = _d.value;
                                /** @type {?} */
                                var parameter = new ActualParameter(process.Model, p.Name, p.Value, p.ValueText, p.Tag || 'expression');
                                action.actualParameters.push(parameter);
                            }
                        }
                        catch (e_29_1) { e_29 = { error: e_29_1 }; }
                        finally {
                            try {
                                if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                            }
                            finally { if (e_29) throw e_29.error; }
                        }
                    }
                    extendElements.push(action);
                }
            }
            catch (e_28_1) { e_28 = { error: e_28_1 }; }
            finally {
                try {
                    if (customizeActions_1_1 && !customizeActions_1_1.done && (_a = customizeActions_1.return)) _a.call(customizeActions_1);
                }
                finally { if (e_28) throw e_28.error; }
            }
        }
    };
    /**
     * 更新流程备注
     * @param bpmnModel
     * @param value
     */
    /**
     * 更新流程备注
     * @param {?} value
     * @return {?}
     */
    WfProcessCell.prototype.modifyProcessTitleExpression = /**
     * 更新流程备注
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        if (extendElements.length > 0) {
            for (var i = extendElements.length; i > 0; i--) {
                if (extendElements[i - 1] instanceof TitleExpression) {
                    extendElements.splice(i - 1, 1);
                }
            }
        }
        /** @type {?} */
        var title = new TitleExpression(process.Model);
        if (typeof (value) === 'string' || value == null) {
            title = new TitleExpression(process.Model, value);
        }
        else {
            title = new TitleExpression(process.Model, value['zh-CHS'], value);
        }
        extendElements.push(title);
    };
    /**
     * @param {?} ffName
     * @param {?=} bizObjId
     * @param {?=} ffId
     * @return {?}
     */
    WfProcessCell.prototype.updateProcessHeader = /**
     * @param {?} ffName
     * @param {?=} bizObjId
     * @param {?=} ffId
     * @return {?}
     */
    function (ffName, bizObjId, ffId) {
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        if (extendElements !== null && extendElements.length > 0) {
            for (var i in extendElements) {
                if (extendElements[i] instanceof ProcessHeader) {
                    ((/** @type {?} */ (extendElements[i]))).BizDefName = ffName;
                    ((/** @type {?} */ (extendElements[i]))).BizObjId = bizObjId;
                    ((/** @type {?} */ (extendElements[i]))).BizDefKey = ffId;
                    return;
                }
            }
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    WfProcessCell.prototype.bindProcessTitleExpression = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        /** @type {?} */
        var titleExpression = (/** @type {?} */ (extendElements.find((/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return e.ClrTypeID === TitleExpression.Clr_Type_ID; }))));
        if (titleExpression && titleExpression.value) {
            return;
        }
        /** @type {?} */
        var title = new TitleExpression(process.Model, value);
        extendElements.push(title);
    };
    /**
     * 基于流程分类的切换更新流程构件
     */
    /**
     * 基于流程分类的切换更新流程构件
     * @param {?} procCpts
     * @return {?}
     */
    WfProcessCell.prototype.updateProcessComponents = /**
     * 基于流程分类的切换更新流程构件
     * @param {?} procCpts
     * @return {?}
     */
    function (procCpts) {
        var e_30, _a, e_31, _b;
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        //删除流程构件
        if (extendElements.length > 0) {
            for (var i = extendElements.length; i > 0; i--) {
                if (extendElements[i - 1] instanceof Component
                    && ((/** @type {?} */ (extendElements[i - 1]))).Time !== ToolTime.UNDEFINITION) {
                    extendElements.splice(i - 1, 1);
                }
            }
        }
        //重新绑定
        if (procCpts && procCpts.length > 0) {
            try {
                for (var procCpts_4 = tslib_1.__values(procCpts), procCpts_4_1 = procCpts_4.next(); !procCpts_4_1.done; procCpts_4_1 = procCpts_4.next()) {
                    var procCpt = procCpts_4_1.value;
                    /** @type {?} */
                    var cpt = new Component(process.Model);
                    cpt.Time = (/** @type {?} */ (procCpt.executionTime));
                    cpt.MetadataId = procCpt.metadataId;
                    cpt.MethodCode = procCpt.methodCode;
                    cpt.Id = procCpt.id;
                    cpt.Name = procCpt.name;
                    cpt.isForCompensation = procCpt.isForCompensation;
                    cpt.ActualParameters = [];
                    if (procCpt.actualParameters && procCpt.actualParameters.length > 0) {
                        try {
                            for (var _c = tslib_1.__values(procCpt.actualParameters), _d = _c.next(); !_d.done; _d = _c.next()) {
                                var item = _d.value;
                                /** @type {?} */
                                var parameter = new ActualParameter(process.Model, item.code, item.value);
                                cpt.ActualParameters.push(parameter);
                            }
                        }
                        catch (e_31_1) { e_31 = { error: e_31_1 }; }
                        finally {
                            try {
                                if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                            }
                            finally { if (e_31) throw e_31.error; }
                        }
                    }
                    extendElements.push(cpt);
                }
            }
            catch (e_30_1) { e_30 = { error: e_30_1 }; }
            finally {
                try {
                    if (procCpts_4_1 && !procCpts_4_1.done && (_a = procCpts_4.return)) _a.call(procCpts_4);
                }
                finally { if (e_30) throw e_30.error; }
            }
        }
    };
    /**
     * @param {?} assign
     * @return {?}
     */
    WfProcessCell.prototype.updateAllowPreAssign = /**
     * @param {?} assign
     * @return {?}
     */
    function (assign) {
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        if (extendElements !== null && extendElements.length > 0) {
            for (var i in extendElements) {
                if (extendElements[i] instanceof ProcessHeader) {
                    ((/** @type {?} */ (extendElements[i]))).allowPreAssign = assign;
                    return;
                }
            }
        }
    };
    /**
     * @param {?} priority
     * @return {?}
     */
    WfProcessCell.prototype.updateEnablePriority = /**
     * @param {?} priority
     * @return {?}
     */
    function (priority) {
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        if (extendElements !== null && extendElements.length > 0) {
            for (var i in extendElements) {
                if (extendElements[i] instanceof ProcessHeader) {
                    ((/** @type {?} */ (extendElements[i]))).enablePriority = priority;
                    return;
                }
            }
        }
    };
    /**
     * @param {?} startMode
     * @param {?} allCmps
     * @param {?} bizCmps
     * @return {?}
     */
    WfProcessCell.prototype.updateStartMode = /**
     * @param {?} startMode
     * @param {?} allCmps
     * @param {?} bizCmps
     * @return {?}
     */
    function (startMode, allCmps, bizCmps) {
        var e_32, _a;
        // 对于启动方式的提示信息
        switch (startMode) {
            case StartMode.DIRECT:
                this.editorUi.fireEvent(new mxEventObject(WfEvents.ShowTips, "data", {
                    type: 'message',
                    level: 'info',
                    message: mxResources.get("directTips"),
                }));
                break;
            case StartMode.EXTERNAL_PROCESS:
                this.editorUi.fireEvent(new mxEventObject(WfEvents.ShowTips, "data", {
                    type: 'message',
                    level: 'info',
                    message: mxResources.get("externalTips"),
                }));
                break;
            case StartMode.SUB_PROCESS:
                this.editorUi.fireEvent(new mxEventObject(WfEvents.ShowTips, "data", {
                    type: 'message',
                    level: 'info',
                    message: mxResources.get("subprocessTips"),
                }));
                break;
        }
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        if (extendElements !== null && extendElements.length > 0) {
            /** @type {?} */
            var procHeader = (/** @type {?} */ (extendElements.filter((/**
             * @param {?} e
             * @return {?}
             */
            function (e) { return e instanceof ProcessHeader; }))));
            // 修改启动方式
            procHeader[0].startMode = startMode;
            /** @type {?} */
            var procCpts = allCmps;
            //删除流程构件
            if (extendElements.length > 0) {
                for (var i = extendElements.length; i > 0; i--) {
                    if (extendElements[i - 1] instanceof Component
                        && ((/** @type {?} */ (extendElements[i - 1]))).Time !== ToolTime.UNDEFINITION
                        && ((/** @type {?} */ (extendElements[i - 1]))).Time !== ToolTime.LOADDATA) {
                        extendElements.splice(i - 1, 1);
                    }
                }
            }
            allCmps && allCmps.forEach((/**
             * @param {?} procCpt
             * @return {?}
             */
            function (procCpt) {
                /** @type {?} */
                var cpt = new Component(process.Model);
                cpt.Time = (/** @type {?} */ (procCpt.executionTime));
                cpt.MetadataId = procCpt.metadataId;
                cpt.MethodCode = procCpt.methodCode;
                cpt.Id = procCpt.id;
                cpt.Name = procCpt.name;
                cpt.isForCompensation = procCpt.isForCompensation;
                cpt.ActualParameters = [];
                if (procCpt.scopeType) {
                    cpt.ScopeType = procCpt.scopeType;
                }
                else {
                    cpt.ScopeType = "Undefined";
                }
                procCpt.actualParameters && procCpt.actualParameters.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    /** @type {?} */
                    var parameter = new ActualParameter(process.Model, item.code, item.value);
                    cpt.ActualParameters.push(parameter);
                }));
                extendElements.push(cpt);
            }));
            /** @type {?} */
            var procCmps = (/** @type {?} */ (extendElements.filter((/**
             * @param {?} e
             * @return {?}
             */
            function (e) { return e instanceof Component; }))));
            // 如果启动方式是子流程,只保留取数构件
            if (startMode === StartMode.SUB_PROCESS) {
                // Utils.postMessage(ResourceKeys.wf_showTip, {
                //     type: 'message',
                //     level: 'info',
                //     message: `启动方式为子流程，需要移除取数构件以外的流程构件`,
                // });
                for (var i = extendElements.length; i > 0; i--) {
                    if (extendElements[i - 1] instanceof Component) {
                        /** @type {?} */
                        var cmp = (/** @type {?} */ (extendElements[i - 1]));
                        if (cmp.Time !== ToolTime.UNDEFINITION
                            && cmp.Time !== ToolTime.LOADDATA
                            && (cmp.ScopeType !== StartMode.SUB_PROCESS && cmp.ScopeType !== "All_Process")) {
                            extendElements.splice(i - 1, 1);
                        }
                    }
                }
            }
            // 如果启动方式是直接启动,不允许存在共享的构件
            if (startMode === StartMode.DIRECT) {
                // 判断有没有共享（PF）的流程构件
                /** @type {?} */
                var pfCmps = procCmps.filter((/**
                 * @param {?} p
                 * @return {?}
                 */
                function (p) { return WfDesignerUtils.isPfComponent(p.MetadataId); }));
                if (pfCmps && pfCmps.length > 0) {
                    // 删掉共享的构件和启动方式
                    for (var i = extendElements.length; i > 0; i--) {
                        if (extendElements[i - 1] instanceof Component) {
                            /** @type {?} */
                            var cmp = (/** @type {?} */ (extendElements[i - 1]));
                            if (WfDesignerUtils.isPfComponent(cmp.MetadataId)
                                || cmp.ScopeType === StartMode.EXTERNAL_PROCESS
                                || cmp.ScopeType === StartMode.SUB_PROCESS) {
                                extendElements.splice(i - 1, 1);
                            }
                        }
                    }
                }
                // 如果启动方式是外部流程
            }
            else if (startMode === StartMode.EXTERNAL_PROCESS) {
                // const otherCmps = procCmps.filter((p: Component) => (p.MetadataId !== ResourceKeys.pf_Cmp_Id1
                //     && p.MetadataId !== ResourceKeys.pf_Cmp_Id2
                //     && p.MetadataId !== ResourceKeys.pf_Cmp_Id3
                //     && p.MetadataId !== ResourceKeys.pf_Cmp_Id4
                //     && p.Time !== ToolTime.LOADDATA
                // ));
                // 判断有没有预制外部流程通用构件
                /** @type {?} */
                var externalCmps = false;
                /** @type {?} */
                var cmps = bizCmps;
                if (cmps && cmps.length > 0) {
                    try {
                        // 此处标记是外部流程通用构件
                        for (var cmps_2 = tslib_1.__values(cmps), cmps_2_1 = cmps_2.next(); !cmps_2_1.done; cmps_2_1 = cmps_2.next()) {
                            var c = cmps_2_1.value;
                            if (c.scopeType === 'ExternalProcess') {
                                externalCmps = true;
                                break;
                            }
                        }
                    }
                    catch (e_32_1) { e_32 = { error: e_32_1 }; }
                    finally {
                        try {
                            if (cmps_2_1 && !cmps_2_1.done && (_a = cmps_2.return)) _a.call(cmps_2);
                        }
                        finally { if (e_32) throw e_32.error; }
                    }
                    // 如果预制了外部流程通用构件,只保留取数构件，没有则不处理
                    if (externalCmps) {
                        for (var i = extendElements.length; i > 0; i--) {
                            if (extendElements[i - 1] instanceof Component) {
                                /** @type {?} */
                                var cmp = (/** @type {?} */ (extendElements[i - 1]));
                                if (cmp.Time !== ToolTime.UNDEFINITION
                                    && cmp.Time !== ToolTime.LOADDATA
                                    && (cmp.ScopeType === StartMode.DIRECT
                                        || cmp.ScopeType === StartMode.SUB_PROCESS
                                        || cmp.ScopeType === "Undefined")) {
                                    extendElements.splice(i - 1, 1);
                                }
                            }
                        }
                    }
                }
            }
        }
        this.editorUi.fireEvent(new mxEventObject(PmfEvents.ShowProperty));
    };
    /**
     * @param {?} variables
     * @return {?}
     */
    WfProcessCell.prototype.updateProcessVariables = /**
     * @param {?} variables
     * @return {?}
     */
    function (variables) {
        var e_33, _a, e_34, _b, e_35, _c;
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var objects = process.GetDataObjects();
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        if (objects && objects.length > 0) {
            var _loop_2 = function (o) {
                if (o instanceof ValuedDataObject && o.DataType.Category !== DataTypes.CustomType) {
                    if (o.variableSetting && o.variableSetting.tag === 'component') {
                        /** @type {?} */
                        var i = extendElements.findIndex((/**
                         * @param {?} e
                         * @return {?}
                         */
                        function (e) { return (e instanceof Component && e.Id === o.variableSetting.value); }));
                        extendElements.splice(i, 1);
                    }
                    process.RemoveDataObject(o.Id);
                }
            };
            try {
                for (var objects_3 = tslib_1.__values(objects), objects_3_1 = objects_3.next(); !objects_3_1.done; objects_3_1 = objects_3.next()) {
                    var o = objects_3_1.value;
                    _loop_2(o);
                }
            }
            catch (e_33_1) { e_33 = { error: e_33_1 }; }
            finally {
                try {
                    if (objects_3_1 && !objects_3_1.done && (_a = objects_3.return)) _a.call(objects_3);
                }
                finally { if (e_33) throw e_33.error; }
            }
        }
        if (variables !== null && variables.length > 0) {
            try {
                for (var variables_1 = tslib_1.__values(variables), variables_1_1 = variables_1.next(); !variables_1_1.done; variables_1_1 = variables_1.next()) {
                    var v = variables_1_1.value;
                    /** @type {?} */
                    var dataType = new DataType(process.Model);
                    dataType.Category = DataTypes.BasicType;
                    dataType.BasicType = v.basicType;
                    dataType.Description = v.description;
                    /** @type {?} */
                    var dataObject = new ValuedDataObject(process.Model);
                    dataObject.Id = v.id;
                    dataObject.Name = v.name;
                    dataObject.Documentation = v.documentation;
                    dataObject.DataType = dataType;
                    if (v.tag && v.value) {
                        /** @type {?} */
                        var variableSetting = new VariableSetting(process.Model);
                        variableSetting.variableCode = v.name;
                        variableSetting.variableName = v.documentation;
                        variableSetting.tag = v.tag;
                        variableSetting.value = v.value;
                        variableSetting.valueText = v.valueText;
                        dataObject.variableSetting = variableSetting;
                        if (variableSetting.tag === 'component') {
                            /** @type {?} */
                            var c = v.component;
                            if (c) {
                                /** @type {?} */
                                var cpt = new Component(process.Model);
                                cpt.MetadataId = c.MetadataId;
                                cpt.MethodCode = c.MethodCode;
                                cpt.Id = c.Id;
                                cpt.Name = c.Name;
                                cpt.ActualParameters = new Array();
                                if (c.ActualParameters && c.ActualParameters.length > 0) {
                                    try {
                                        for (var _d = tslib_1.__values(c.ActualParameters), _e = _d.next(); !_e.done; _e = _d.next()) {
                                            var a = _e.value;
                                            /** @type {?} */
                                            var parameter = new ActualParameter(process.Model, a.Name, a.Value);
                                            parameter.Tag = a.Tag;
                                            cpt.ActualParameters.push(parameter);
                                        }
                                    }
                                    catch (e_35_1) { e_35 = { error: e_35_1 }; }
                                    finally {
                                        try {
                                            if (_e && !_e.done && (_c = _d.return)) _c.call(_d);
                                        }
                                        finally { if (e_35) throw e_35.error; }
                                    }
                                }
                                extendElements.push(cpt);
                            }
                        }
                    }
                    process.AddDataObject(dataObject);
                }
            }
            catch (e_34_1) { e_34 = { error: e_34_1 }; }
            finally {
                try {
                    if (variables_1_1 && !variables_1_1.done && (_b = variables_1.return)) _b.call(variables_1);
                }
                finally { if (e_34) throw e_34.error; }
            }
        }
    };
    /**
     * @param {?} timeLimit
     * @return {?}
     */
    WfProcessCell.prototype.updateProcessTimeLimitControl = /**
     * @param {?} timeLimit
     * @return {?}
     */
    function (timeLimit) {
        var e_36, _a, e_37, _b;
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        if (extendElements.length > 0) {
            for (var i = extendElements.length; i > 0; i--) {
                if (extendElements[i - 1] instanceof TimeLimitControl) {
                    extendElements.splice(i - 1, 1);
                }
            }
        }
        /** @type {?} */
        var timeLimitInfo = (/** @type {?} */ (timeLimit));
        if (timeLimitInfo) {
            /** @type {?} */
            var control = new TimeLimitControl(process.Model);
            control.enable = timeLimitInfo.enable;
            control.tag = timeLimitInfo.tag;
            control.dueTime = timeLimitInfo.dueTime;
            control.timeUnit = timeLimitInfo.timeUnit;
            control.expireOperations = new Array();
            if (timeLimitInfo.expireOperations && timeLimitInfo.expireOperations.length > 0) {
                try {
                    for (var _c = tslib_1.__values(timeLimitInfo.expireOperations), _d = _c.next(); !_d.done; _d = _c.next()) {
                        var o = _d.value;
                        if (o.operationTypeName === 'SendMessageExpireOperation' || !o.operationTypeName) {
                            /** @type {?} */
                            var operation = (/** @type {?} */ (o));
                            /** @type {?} */
                            var sendMessage = new SendMessageExpireOperation(process.Model);
                            sendMessage.id = BpmnModelHelper.GenerateElementId();
                            sendMessage.durationType = (/** @type {?} */ (operation.durationType));
                            sendMessage.timeDuration = operation.timeDuration;
                            sendMessage.timeUnit = operation.timeUnit;
                            sendMessage.operationTypeName = operation.operationTypeName ? operation.operationTypeName : 'SendMessageExpireOperation';
                            sendMessage.message = operation.message;
                            sendMessage.subject = operation.subject;
                            sendMessage.sendType = operation.sendType;
                            sendMessage.receiver = operation.receiver;
                            sendMessage.receiverText = operation.receiverText;
                            sendMessage.cycleEnable = operation.cycleEnable;
                            sendMessage.cycleTime = operation.cycleTime;
                            sendMessage.cycleTimeUnit = operation.cycleTimeUnit;
                            if (operation.smsMessageTemplate) {
                                /** @type {?} */
                                var smsMessageTemplate = new SmsMessageTemplate(process.Model);
                                smsMessageTemplate.code = operation.smsMessageTemplate.code;
                                smsMessageTemplate.content = operation.smsMessageTemplate.content;
                                smsMessageTemplate.smsMessageVariables = new Array();
                                if (operation.smsMessageTemplate.smsMessageVariables && operation.smsMessageTemplate.smsMessageVariables.length > 0) {
                                    try {
                                        for (var _e = tslib_1.__values(operation.smsMessageTemplate.smsMessageVariables), _f = _e.next(); !_f.done; _f = _e.next()) {
                                            var v = _f.value;
                                            /** @type {?} */
                                            var variable = new SmsMessageVariable(process.Model);
                                            variable.code = v.code;
                                            variable.value = v.value;
                                            variable.tag = v.tag;
                                            smsMessageTemplate.smsMessageVariables.push(variable);
                                        }
                                    }
                                    catch (e_37_1) { e_37 = { error: e_37_1 }; }
                                    finally {
                                        try {
                                            if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                                        }
                                        finally { if (e_37) throw e_37.error; }
                                    }
                                }
                                sendMessage.smsMessageTemplate = smsMessageTemplate;
                            }
                            control.expireOperations.push(sendMessage);
                        }
                        else if (o.operationTypeName === 'AutoHandleExpireOperation') {
                            /** @type {?} */
                            var operation = (/** @type {?} */ (o));
                            /** @type {?} */
                            var autoHandle = new AutoHandleExpireOperation(process.Model);
                            autoHandle.id = BpmnModelHelper.GenerateElementId();
                            autoHandle.durationType = (/** @type {?} */ (operation.durationType));
                            autoHandle.timeDuration = operation.timeDuration;
                            autoHandle.timeUnit = operation.timeUnit;
                            autoHandle.operationTypeName = operation.operationTypeName ? operation.operationTypeName : 'AutoHandleExpireOperation';
                            autoHandle.autoHandleType = operation.autoHandleType;
                            autoHandle.message = operation.message;
                            control.expireOperations.push(autoHandle);
                        }
                    }
                }
                catch (e_36_1) { e_36 = { error: e_36_1 }; }
                finally {
                    try {
                        if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                    }
                    finally { if (e_36) throw e_36.error; }
                }
            }
            extendElements.push(control);
        }
    };
    /**
     * @param {?} notices
     * @return {?}
     */
    WfProcessCell.prototype.updateProcessMessageNotices = /**
     * @param {?} notices
     * @return {?}
     */
    function (notices) {
        var e_38, _a, e_39, _b, e_40, _c, e_41, _d;
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        if (extendElements.length > 0) {
            for (var i = extendElements.length; i > 0; i--) {
                if (extendElements[i - 1] instanceof MessageNotice) {
                    extendElements.splice(i - 1, 1);
                }
            }
        }
        if (notices && notices.length > 0) {
            try {
                for (var notices_1 = tslib_1.__values(notices), notices_1_1 = notices_1.next(); !notices_1_1.done; notices_1_1 = notices_1.next()) {
                    var n = notices_1_1.value;
                    /** @type {?} */
                    var messageNotice = new MessageNotice(process.Model);
                    messageNotice.sendType = n.sendType;
                    messageNotice.sendTime = n.sendTime;
                    messageNotice.subject = n.subject;
                    messageNotice.content = n.content;
                    // messageNotice.receiver = n.receiver;
                    // messageNotice.receiverText = n.receiverText;
                    n.receivers = n.receivers.filter((/**
                     * @param {?} r
                     * @return {?}
                     */
                    function (r) { return r.type; }));
                    if (n.receivers && n.receivers.length > 0) {
                        /** @type {?} */
                        var receivers = n.receivers.filter((/**
                         * @param {?} r
                         * @return {?}
                         */
                        function (r) { return r.type === 'userParticipant'
                            || (r.type === 'processParticipant' && r.processParticipant !== 'activityExecutorParticipant'); }));
                        /** @type {?} */
                        var participants = n.receivers.filter((/**
                         * @param {?} r
                         * @return {?}
                         */
                        function (r) { return r.type === 'variableParticipant' || r.type === 'positionParticipant'
                            || (r.type === 'processParticipant' && r.processParticipant === 'activityExecutorParticipant'); }));
                        if (receivers && receivers.length > 0) {
                            /** @type {?} */
                            var receiver = '';
                            /** @type {?} */
                            var receiverName = '';
                            try {
                                for (var receivers_1 = tslib_1.__values(receivers), receivers_1_1 = receivers_1.next(); !receivers_1_1.done; receivers_1_1 = receivers_1.next()) {
                                    var r = receivers_1_1.value;
                                    if (r.type === 'processParticipant') {
                                        receiver += (r.processParticipant + ',');
                                        if (r.processParticipant === 'procInstCreator') {
                                            receiverName += '流程发起人,';
                                        }
                                        else if (r.processParticipant === 'workItemParticipant') {
                                            receiverName += '任务接收人,';
                                        }
                                    }
                                    else if (r.type === 'userParticipant') {
                                        receiver += (r.userIds + ',');
                                        receiverName += (r.userNames + ',');
                                    }
                                }
                            }
                            catch (e_39_1) { e_39 = { error: e_39_1 }; }
                            finally {
                                try {
                                    if (receivers_1_1 && !receivers_1_1.done && (_b = receivers_1.return)) _b.call(receivers_1);
                                }
                                finally { if (e_39) throw e_39.error; }
                            }
                            if (receiver) {
                                receiver = receiver.substring(0, receiver.length - 1);
                            }
                            if (receiverName) {
                                receiverName = receiverName.substring(0, receiverName.length - 1);
                            }
                            messageNotice.receiver = receiver;
                            messageNotice.receiverText = receiverName;
                        }
                        if (participants && participants.length > 0) {
                            /** @type {?} */
                            var receiverExpressions = new Array();
                            try {
                                for (var participants_1 = tslib_1.__values(participants), participants_1_1 = participants_1.next(); !participants_1_1.done; participants_1_1 = participants_1.next()) {
                                    var p = participants_1_1.value;
                                    if (p.type === 'processParticipant') {
                                        /** @type {?} */
                                        var activityExecutor = new ActivityExecutorParticipantExpression(process.Model);
                                        activityExecutor.id = BpmnModelHelper.GenerateElementId();
                                        activityExecutor.activityDefId = p.activityDefId;
                                        activityExecutor.activityDefName = p.activityDefName;
                                        receiverExpressions.push(activityExecutor);
                                    }
                                    if (p.type === 'variableParticipant') {
                                        /** @type {?} */
                                        var variableParticipant = new UserParticipantExpression(process.Model, p.variableParticipant, p.variableName, p.orgType);
                                        variableParticipant.id = BpmnModelHelper.GenerateElementId();
                                        variableParticipant.Tag = 'expression';
                                        variableParticipant.OrgType = p.orgType;
                                        receiverExpressions.push(variableParticipant);
                                    }
                                    if (p.type === 'positionParticipant') {
                                        /** @type {?} */
                                        var posiParticipant = new PositionParticipantExpression(process.Model, p.positionId, p.positionName);
                                        posiParticipant.id = BpmnModelHelper.GenerateElementId();
                                        posiParticipant.PositionType = p.positionType;
                                        if (p.positionType === '1') {
                                            posiParticipant.OrgId = p.orgId;
                                            posiParticipant.OrgType = p.orgType;
                                        }
                                        receiverExpressions.push(posiParticipant);
                                    }
                                }
                            }
                            catch (e_40_1) { e_40 = { error: e_40_1 }; }
                            finally {
                                try {
                                    if (participants_1_1 && !participants_1_1.done && (_c = participants_1.return)) _c.call(participants_1);
                                }
                                finally { if (e_40) throw e_40.error; }
                            }
                            messageNotice.receiverExpressions = receiverExpressions;
                        }
                    }
                    if (n.smsMessageTemplate) {
                        /** @type {?} */
                        var smsMessageTemplate = new SmsMessageTemplate(process.Model);
                        smsMessageTemplate.code = n.smsMessageTemplate.code;
                        smsMessageTemplate.content = n.smsMessageTemplate.content;
                        smsMessageTemplate.smsMessageVariables = new Array();
                        if (n.smsMessageTemplate.smsMessageVariables && n.smsMessageTemplate.smsMessageVariables.length > 0) {
                            try {
                                for (var _e = tslib_1.__values(n.smsMessageTemplate.smsMessageVariables), _f = _e.next(); !_f.done; _f = _e.next()) {
                                    var v = _f.value;
                                    /** @type {?} */
                                    var variable = new SmsMessageVariable(process.Model);
                                    variable.code = v.code;
                                    variable.value = v.value;
                                    variable.tag = v.tag;
                                    smsMessageTemplate.smsMessageVariables.push(variable);
                                }
                            }
                            catch (e_41_1) { e_41 = { error: e_41_1 }; }
                            finally {
                                try {
                                    if (_f && !_f.done && (_d = _e.return)) _d.call(_e);
                                }
                                finally { if (e_41) throw e_41.error; }
                            }
                        }
                        messageNotice.smsMessageTemplate = smsMessageTemplate;
                    }
                    extendElements.push(messageNotice);
                }
            }
            catch (e_38_1) { e_38 = { error: e_38_1 }; }
            finally {
                try {
                    if (notices_1_1 && !notices_1_1.done && (_a = notices_1.return)) _a.call(notices_1);
                }
                finally { if (e_38) throw e_38.error; }
            }
        }
    };
    /**
     * @param {?} imageSettings
     * @return {?}
     */
    WfProcessCell.prototype.updateImageSettings = /**
     * @param {?} imageSettings
     * @return {?}
     */
    function (imageSettings) {
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        if (extendElements.length > 0) {
            for (var i = extendElements.length; i > 0; i--) {
                if (extendElements[i - 1] instanceof ImageSettings) {
                    extendElements.splice(i - 1, 1);
                }
            }
        }
        if (imageSettings) {
            /** @type {?} */
            var imageSetting = new ImageSettings(process.Model);
            imageSetting.imageOperation = imageSettings.imageOperation;
            extendElements.push(imageSetting);
        }
    };
    /**
     * 更新扩展属性
     * @param bpmnModel
     * @param value
     */
    /**
     * 更新扩展属性
     * @param {?} value
     * @return {?}
     */
    WfProcessCell.prototype.resetExtendProperty = /**
     * 更新扩展属性
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var e_42, _a, e_43, _b;
        /** @type {?} */
        var process = this.process;
        /** @type {?} */
        var extendElements = process.GetExtensionElements();
        try {
            for (var value_1 = tslib_1.__values(value), value_1_1 = value_1.next(); !value_1_1.done; value_1_1 = value_1.next()) {
                var o = value_1_1.value;
                try {
                    for (var extendElements_2 = tslib_1.__values(extendElements), extendElements_2_1 = extendElements_2.next(); !extendElements_2_1.done; extendElements_2_1 = extendElements_2.next()) {
                        var e = extendElements_2_1.value;
                        if (e instanceof ExtendProperty && e.key === o.code) {
                            /** @type {?} */
                            var index = extendElements.indexOf(e);
                            if (index > -1) {
                                extendElements.splice(index, 1);
                            }
                        }
                    }
                }
                catch (e_43_1) { e_43 = { error: e_43_1 }; }
                finally {
                    try {
                        if (extendElements_2_1 && !extendElements_2_1.done && (_b = extendElements_2.return)) _b.call(extendElements_2);
                    }
                    finally { if (e_43) throw e_43.error; }
                }
            }
        }
        catch (e_42_1) { e_42 = { error: e_42_1 }; }
        finally {
            try {
                if (value_1_1 && !value_1_1.done && (_a = value_1.return)) _a.call(value_1);
            }
            finally { if (e_42) throw e_42.error; }
        }
        /** @type {?} */
        var attributes = (/** @type {?} */ (value));
        if (attributes) {
            attributes.forEach((/**
             * @param {?} a
             * @return {?}
             */
            function (a) {
                /** @type {?} */
                var item = new ExtendProperty(process.Model);
                item.key = a.code;
                item.value = a.value;
                extendElements.push(item);
            }));
        }
    };
    /**
     * @param {?} wholeOrderRetrieve
     * @return {?}
     */
    WfProcessCell.prototype.updateEnableWholeOrderRetrieve = /**
     * @param {?} wholeOrderRetrieve
     * @return {?}
     */
    function (wholeOrderRetrieve) {
        /** @type {?} */
        var extendElements = this.process.GetExtensionElements();
        if (extendElements !== null && extendElements.length > 0) {
            for (var i in extendElements) {
                if (extendElements[i] instanceof ProcessHeader) {
                    ((/** @type {?} */ (extendElements[i]))).enableWholeOrderRetrieve = wholeOrderRetrieve;
                    return;
                }
            }
        }
    };
    return WfProcessCell;
}(ProcessCell));
/**
 * 流程元素类。
 * 取名虽然Cell，但实际上不需要继承mxCell以及Cell类，仅持有流程与画布实例，方法属性获取与修改操作。
 */
export { WfProcessCell };
if (false) {
    /** @type {?} */
    WfProcessCell.prototype.process;
    /** @type {?} */
    WfProcessCell.prototype.plane;
    /** @type {?} */
    WfProcessCell.prototype.editorUi;
    /** @type {?} */
    WfProcessCell.prototype.uiState;
}
