// @gsp-cmp/ubcc-param-vo/esm2015/lib/vo/serializer/data-type-serializer/help-type-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/data-type-serializer/help-type-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DefaultStructuredTypeSerializer } from '@ecp-caf/common-structure';
import { HelpType } from '../../extension/data-type/help-type';
import { SerializerUtils } from '../../utils/serializer-utils';
export class HelpTypeSerialzier {
    /**
     * @param {?} type
     * @param {?} ctxt
     * @return {?}
     */
    serialize(type, ctxt) {
        /** @type {?} */
        let jobj = new Object();
        if (!type) {
            return null;
        }
        /** @type {?} */
        let serializer = new DefaultStructuredTypeSerializer();
        jobj = serializer.serialize(type, SerializerUtils.getContext());
        /** @type {?} */
        let helpType = (/** @type {?} */ (type));
        jobj['helpId'] = helpType.helpId;
        jobj['helpCode'] = helpType.helpCode;
        jobj['helpName'] = helpType.helpName;
        jobj['kind'] = helpType.kind;
        return jobj;
    }
    /**
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    deserialize(obj, ctxt) {
        if (!obj) {
            return null;
        }
        /** @type {?} */
        let helpType = new HelpType();
        /** @type {?} */
        let serializer = new DefaultStructuredTypeSerializer();
        helpType = (/** @type {?} */ (serializer.deserialize(obj, SerializerUtils.getContext())));
        helpType.helpId = obj['helpId'];
        helpType.helpCode = obj['helpCode'];
        helpType.helpName = obj['helpName'];
        return helpType;
    }
}
