// @gsp-cmp/ubcc-param-vo/esm2015/lib/vo/serializer/data-type-serializer/structure-type-ref-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/data-type-serializer/structure-type-ref-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { StructureTypeRef } from '../../extension/data-type/structure-type-ref';
export class StructureTypeRefSerializer {
    /**
     * @param {?} type
     * @param {?} ctxt
     * @return {?}
     */
    serialize(type, ctxt) {
        if (!type) {
            return null;
        }
        /** @type {?} */
        let jobj = new Object();
        /** @type {?} */
        let refType = (/** @type {?} */ (type));
        jobj['kind'] = refType.kind;
        jobj['refId'] = refType.refId;
        jobj['type'] = refType.type;
        return jobj;
    }
    /**
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    deserialize(obj, ctxt) {
        if (!obj) {
            return null;
        }
        /** @type {?} */
        let refType = new StructureTypeRef();
        refType.refId = obj['refId'];
        refType.type = (/** @type {?} */ (obj['type']));
        return refType;
    }
}
