// @farris/ui-search-box/esm5/lib/search-fields/editors/boolean/search-editor-boolean.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, ChangeDetectorRef, Component, ElementRef, Injector, Input, Output } from '@angular/core';
var SearchFieldEditorBooleanComponent = /** @class */ (function () {
    function SearchFieldEditorBooleanComponent(injector, cd, el) {
        this.injector = injector;
        this.cd = cd;
        this.el = el;
        this.radioDataSource = [];
        this.selectedValue = undefined;
        this.valueChange = new EventEmitter();
        this._radioElements = [];
    }
    /**
     * @return {?}
     */
    SearchFieldEditorBooleanComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.field) {
            if (this.field.control) {
                this.radioDataSource = this.field.control.enumValues || [];
            }
            if (this.field.value) {
                this.selectedValue = this.field.value.value;
            }
        }
    };
    /**
     * @return {?}
     */
    SearchFieldEditorBooleanComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.selectedValue !== null && this.selectedValue !== undefined) {
            this._radioElements = Array.from(this.el.nativeElement.querySelectorAll('input'));
            /** @type {?} */
            var radioEl = this._radioElements.find((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return n.value == _this.selectedValue; }));
            if (radioEl) {
                radioEl.checked = true;
            }
        }
    };
    /**
     * @param {?} $event
     * @param {?} value
     * @return {?}
     */
    SearchFieldEditorBooleanComponent.prototype.onRadioClicked = /**
     * @param {?} $event
     * @param {?} value
     * @return {?}
     */
    function ($event, value) {
        this.selectedValue = value;
        /** @type {?} */
        var radioInputRel = $event.target.querySelector('input');
        // radioInputRel && radioInputRel.setAttribute('checked', true);
        if (radioInputRel) {
            this.clearChecked();
            radioInputRel.checked = true;
            this.cd.detectChanges();
        }
        this.valueChange.emit({ field: this.field.code, value: { value: this.selectedValue, single: true } });
    };
    /**
     * @private
     * @return {?}
     */
    SearchFieldEditorBooleanComponent.prototype.clearChecked = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this._radioElements || !this._radioElements.length) {
            this._radioElements = Array.from(this.el.nativeElement.querySelectorAll('input'));
        }
        this._radioElements.forEach((/**
         * @param {?} r
         * @return {?}
         */
        function (r) {
            // r.removeAttribute('checked');
            r.checked = false;
        }));
    };
    SearchFieldEditorBooleanComponent.decorators = [
        { type: Component, args: [{
                    selector: 'search-field-editor-bool',
                    template: "<div class=\"f-filter-panel-radiogroup\">\r\n    <div class=\"panel-radiogroup-item\" \r\n    [ngClass]=\"{'panel-radiogroup-item-active': selectedValue == item.value}\"\r\n    *ngFor=\"let item of radioDataSource\"\r\n    (click)=\"onRadioClicked($event, item.value)\"\r\n    >\r\n        \r\n\r\n        <div class=\"custom-control custom-radio\">\r\n            <input class=\"custom-control-input\" name=\"{{field.code}}\" id=\"{{'search_radio_' + item.value}}\" type=\"radio\" [attr.value]=\"item.value\">\r\n            <label class=\"custom-control-label\" for=\"{{'search_radio_' + item.value}}\">{{item.name}}</label>\r\n        </div>\r\n\r\n\r\n    </div>\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n</div>"
                }] }
    ];
    /** @nocollapse */
    SearchFieldEditorBooleanComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ChangeDetectorRef },
        { type: ElementRef }
    ]; };
    SearchFieldEditorBooleanComponent.propDecorators = {
        field: [{ type: Input }],
        valueChange: [{ type: Output }]
    };
    return SearchFieldEditorBooleanComponent;
}());
export { SearchFieldEditorBooleanComponent };
if (false) {
    /** @type {?} */
    SearchFieldEditorBooleanComponent.prototype.field;
    /** @type {?} */
    SearchFieldEditorBooleanComponent.prototype.radioDataSource;
    /** @type {?} */
    SearchFieldEditorBooleanComponent.prototype.selectedValue;
    /** @type {?} */
    SearchFieldEditorBooleanComponent.prototype.valueChange;
    /**
     * @type {?}
     * @private
     */
    SearchFieldEditorBooleanComponent.prototype._radioElements;
    /**
     * @type {?}
     * @private
     */
    SearchFieldEditorBooleanComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    SearchFieldEditorBooleanComponent.prototype.cd;
    /**
     * @type {?}
     * @private
     */
    SearchFieldEditorBooleanComponent.prototype.el;
}
