// @gsp-lcm/metadatart-selector4biztype/esm5/metadata/metadatart-select4biztype/metadatainfo.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadatart-select4biztype/metadatainfo.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { GSPMetadataRTService4BizType } from '../metadata.rtservice4biztype';
var MetadataInfoService = /** @class */ (function () {
    function MetadataInfoService(http, gspMetadataService) {
        this.http = http;
        this.gspMetadataService = gspMetadataService;
        // tslint:disable-next-line:variable-name
        this.BusinessObject_Base_Path = '/api/dev/main/v1.0/business-objects';
        this.metadatas = [];
    }
    /**
     * @param {?} metadata
     * @return {?}
     */
    MetadataInfoService.prototype.addMetadataRT = /**
     * @param {?} metadata
     * @return {?}
     */
    function (metadata) {
        this.metadatas.push(metadata);
    };
    /**
     * @return {?}
     */
    MetadataInfoService.prototype.clearMetadataRT = /**
     * @return {?}
     */
    function () {
        this.metadatas = [];
    };
    // 获取包源
    // 获取包源
    /**
     * @param {?} obId
     * @return {?}
     */
    MetadataInfoService.prototype.getBusinessObject = 
    // 获取包源
    /**
     * @param {?} obId
     * @return {?}
     */
    function (obId) {
        /** @type {?} */
        var url = this.BusinessObject_Base_Path + '/' + obId;
        return this.http.get(url);
    };
    /**
     * @return {?}
     */
    MetadataInfoService.prototype.GetMetadataTypeList = /**
     * @return {?}
     */
    function () {
        return this.gspMetadataService.GetMetadataTypeList();
    };
    /**
     * @param {?} typeList
     * @return {?}
     */
    MetadataInfoService.prototype.GetMetadataRTList = /**
     * @param {?} typeList
     * @return {?}
     */
    function (typeList) {
        return this.gspMetadataService.GetAllOriginMdRTListWithType(typeList);
    };
    // 获根据Id取指定元数据
    // 获根据Id取指定元数据
    /**
     * @param {?} metadataID
     * @return {?}
     */
    MetadataInfoService.prototype.GetMetadataRT = 
    // 获根据Id取指定元数据
    /**
     * @param {?} metadataID
     * @return {?}
     */
    function (metadataID) {
        return this.gspMetadataService.GetMetadataRT(metadataID);
    };
    MetadataInfoService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MetadataInfoService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: GSPMetadataRTService4BizType }
    ]; };
    return MetadataInfoService;
}());
export { MetadataInfoService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MetadataInfoService.prototype.BusinessObject_Base_Path;
    /** @type {?} */
    MetadataInfoService.prototype.metadatas;
    /**
     * @type {?}
     * @private
     */
    MetadataInfoService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    MetadataInfoService.prototype.gspMetadataService;
}
