// @farris/command-services/esm2015/lib/command-service.js
import { Injectable } from '@angular/core';
import { filter, take } from 'rxjs/operators';
import { AppContext, ViewModel } from '@farris/devkit';
/**
 * 命令服务
 * @scope FrameConmponent
 */
class CommandService {
    constructor(viewModel, appContext) {
        this.viewModel = viewModel;
        this.appContext = appContext;
    }
    /**
     * context取值方法
     */
    get context() {
        return this.innerContext;
    }
    /**
     * context设值方法
     */
    set context(context) {
        this.innerContext = context;
    }
    /**
     * 执行命令
     * @param commandName 命令名称
     * @param frameId 框架ID
     */
    execute(commandName, frameId) {
        if (!commandName || commandName === '' || commandName === 'undefined') {
            return;
        }
        let viewModel = this.viewModel;
        if (frameId) {
            const targetContext = this.appContext.frameContextManager.getFrameContextById(frameId);
            viewModel = targetContext.viewModel;
        }
        return viewModel[commandName]();
    }
    /**
     * 等待BeSession创建
     * @returns 最新创建出来的BeSession
     */
    waitForBeSession() {
        const frameContext = this.context.frameContext;
        const befRepository = frameContext.repository;
        const sessionService = befRepository.restService.sessionService;
        return sessionService.getBeSessionId().pipe(filter((beSessionId) => {
            return beSessionId !== null;
        }), take(1));
    }
    /**
     * 提取某个任务的执行结果
     * @param taskName 任务名称
     * @reutrn 返回taskName对应任务的执行结果
     */
    extractResult(taskName) {
        const results = this.context.results;
        return results[taskName];
    }
    /**
     * 暂停当前组件的行切换事件抛出
     */
    suspendFrameContextRowSelectedEvent(frameId) {
        let frameContext = this.viewModel.frameContext;
        if (frameId) {
            frameContext = this.viewModel.frameContext.appContext.frameContextManager.getFrameContextById(frameId);
        }
        if (frameContext) {
            frameContext.bindingData.rowSelectedEventSuspend = true;
        }
    }
    resumeFrameContextRowSelectedEvent(frameId) {
        let frameContext = this.viewModel.frameContext;
        if (frameId) {
            frameContext = this.viewModel.frameContext.appContext.frameContextManager.getFrameContextById(frameId);
        }
        if (frameContext) {
            frameContext.bindingData.rowSelectedEventSuspend = false;
        }
    }
}
CommandService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CommandService.ctorParameters = () => [
    { type: ViewModel },
    { type: AppContext }
];
export { CommandService };
