// @farris/command-services/esm2015/lib/data-check.service.js
import { Injectable, Optional } from '@angular/core';
import { empty } from 'rxjs';
import { FormMessageService } from './form-message.service';
import { FormNotifyService } from './form-notify.service';
import { LanguageService } from './languag.service';
/**
 * 数据检查服务
 */
class DataCheckService {
    constructor(messageService, formNotifyService, languageService) {
        this.messageService = messageService;
        this.formNotifyService = formNotifyService;
        this.languageService = languageService;
        if (!this.languageService) {
            this.languageService = LanguageService.getInstance();
        }
    }
    /**
     * 编辑前检查
     */
    checkBeforeEdit(idToEdit) {
        const ifSkipCheck = this.ifSkipCheck(idToEdit);
        if (ifSkipCheck === true) {
            return true;
        }
        if (idToEdit) {
            return true;
        }
        this.formNotifyService.warning(this.languageService.plsSelectEditData, { hideTitle: true });
        return empty();
    }
    /**
     * 查看前检查
     */
    checkBeforeView(idToView) {
        const ifSkipCheck = this.ifSkipCheck(idToView);
        if (ifSkipCheck === true) {
            return true;
        }
        if (idToView) {
            return true;
        }
        this.formNotifyService.warning(this.languageService.plsSelectViewData, { hideTitle: true });
        return empty();
    }
    /**
     * 老表单跳过检查
     * @todo：待删除，兼容老表单
     */
    ifSkipCheck(id) {
        const context = this['context'];
        const params = context.command.params;
        // 表单没有重新保存过的情况
        if (params.hasOwnProperty('idToEdit') === false && params.hasOwnProperty('idToView') === false) {
            return true;
        }
        // 保存过但没有填参数的情况
        // @todo: 会漏掉空串的情况，目前主键解析的返回值是undefined，暂不考虑空串场景
        if (id === '') {
            return true;
        }
        return false;
    }
}
DataCheckService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DataCheckService.ctorParameters = () => [
    { type: FormMessageService },
    { type: FormNotifyService },
    { type: LanguageService, decorators: [{ type: Optional }] }
];
export { DataCheckService };
