// @farris/command-services/esm2015/lib/expression_service.js
import { Injectable, Injector } from "@angular/core";
import { BigNumber } from 'bignumber.js';
import { AppContext, FrameContext, Repository, ENTITY_TEMPLATE, ResolveService, ExpressionUtil, ExpressionExecutor } from "@farris/devkit";
import { of } from "rxjs";
export class ExpressionService {
    constructor(injector, resolveService, frameContext, expressionExecutor) {
        this.injector = injector;
        this.resolveService = resolveService;
        this.frameContext = frameContext;
        this.expressionExecutor = expressionExecutor;
    }
    /**
     * 执行表达式计算
     * @param expression 表达式
     * @param customContext 自定义上下文
     * @returns
     */
    execute(expression, customContext) {
        const deps = this.resolveService.resolve(expression);
        const groupDependencies = ExpressionUtil.getGroupFunctionDependency(expression, this.frameContext.repository.entityTypeInfo);
        const entityContext = this.buildEntityContext(deps, groupDependencies);
        const stateContext = this.buildStateContext();
        const context = Object.assign({ [this.entityOriginalNodeCode]: entityContext }, stateContext, { BigNumber, frameContext: this.frameContext, bindingData: this.frameContext.bindingData, repository: this.frameContext.repository }, customContext);
        return this.expressionExecutor.eval(expression, context);
    }
    /**
     * 执行表达式（返回可观察对象）
     * @param expression 表达式
     * @param customContext 自定义上下文
     * @returns
     */
    executeAsync(expression, customContext) {
        const result = this.execute(expression, customContext);
        return of(result);
    }
    /**
     * 构造实体上下文
     * @param deps
     * @param groupDependencies
     * @param context
     * @returns
     */
    buildEntityContext(deps, groupDependencies, context) {
        let isGroupdMainEntity = false;
        deps.forEach((dep) => {
            const isEntityDependency = this.isEntityDependency(dep);
            const isGroupDependency = groupDependencies.findIndex(item => item === dep) !== -1;
            // 如果依赖的是state，无需处理，现在需要确定的是返回多少实体的问题，和state没有关系
            // 表达式依赖了实体
            if (isEntityDependency) {
                // 是聚合依赖
                if (isGroupDependency) {
                    const dependencyLength = dep.split('/').filter(p => p).length - 1;
                    if (dependencyLength === 1) {
                        // 聚合了主表字段，所有主表数据都需要参与运算，此时已经确定计算的实体上下文了。
                        isGroupdMainEntity = true;
                    }
                    else {
                        // 聚合了子表字段，只需要传递当前实体
                    }
                }
                else {
                    // 当前依赖不是聚合，只需要传递当前实体
                }
            }
        });
        const data = this.getEntity();
        if (isGroupdMainEntity) {
            const collection = this.frameContext.repository.entityCollection.toJSON();
            data['__type__'] = 'List';
            data['__items__'] = collection;
        }
        return data;
    }
    /**
     * 是否为实体依赖
     * @param dep
     * @returns
     */
    isEntityDependency(dep) {
        return dep.startsWith(ENTITY_TEMPLATE);
    }
    /**
     * 获取实体
     * @param event
     * @returns
     */
    getEntity() {
        const entityTypeInfo = this.frameContext.repository.entityTypeInfo;
        const bindingData = this.frameContext.bindingData;
        const childrenEntityPaths = [];
        ExpressionUtil.getChildrenEntityPaths(entityTypeInfo, childrenEntityPaths);
        const entity = this.frameContext.bindingData.list.currentItem.toJSON();
        entity['__type__'] = 'Entity';
        if (!childrenEntityPaths || childrenEntityPaths.length < 1) {
            return entity;
        }
        // 找到所有子表
        childrenEntityPaths.forEach((paths) => {
            const row = ExpressionUtil.getCurrentRowByPaths(paths, bindingData);
            if (row) {
                const propertyName = paths.pop();
                let parent = paths.reduce((object, path) => {
                    return object && object[path] || null;
                }, entity);
                const node = Object.assign({ __items__: [...parent[propertyName]] }, row, { __type__: 'List' });
                parent[propertyName] = node;
            }
        });
        return entity;
    }
    /**
     * 获取主实体原始字段名
     */
    get entityOriginalNodeCode() {
        const repository = this.injector.get(Repository);
        return repository && repository.entityTypeInfo && repository.entityTypeInfo.entityInfo && repository.entityTypeInfo.entityInfo.originalCode || null;
    }
    /**
     * 构造变量上下文
     * @param event
     * @returns
     */
    buildStateContext() {
        const ns = this.frameContext.namespace;
        const appContext = this.injector.get(AppContext, null);
        const frameContexts = appContext.frameContextManager.getFrameContextsByNamespace(ns);
        const result = {};
        if (frameContexts && frameContexts.length > 0) {
            const anonymousFrameContext = frameContexts[0];
            const rootFrameContext = anonymousFrameContext.getVirtualRootFrameContext();
            if (rootFrameContext) {
                const uiState = rootFrameContext.viewModel.uiState;
                const propertyNames = Object.getOwnPropertyNames(uiState) || [];
                propertyNames.forEach((prop) => {
                    if (prop.match(/^[a-zA-Z0-9_\$]+$/g) !== null) {
                        result[prop] = uiState[prop];
                    }
                });
            }
        }
        return result;
    }
}
ExpressionService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ExpressionService.ctorParameters = () => [
    { type: Injector },
    { type: ResolveService },
    { type: FrameContext },
    { type: ExpressionExecutor }
];
