// @farris/command-services/esm2015/lib/form-message.service.js
import { Inject, Injectable, LOCALE_ID, Optional } from '@angular/core';
import { from } from 'rxjs';
import { MessagerService } from '@farris/ui-messager';
import { LanguageService } from './languag.service';
import { ExceptionFactory } from './error/index';
/**
 * 表单消息服务
 * @scope FormModule
 * @summary
 * 1、包装@farris/ui的消息服务；
 * 2、提供针对表单的快捷方法；
 */
class FormMessageService {
    /**
     * 构造函数
     * 注入@farris/ui的LoadingService
     */
    constructor(messagerService, languageService, curLanguage) {
        this.messagerService = messagerService;
        this.languageService = languageService;
        this.curLanguage = curLanguage;
        this.languageService = this.languageService || LanguageService.getInstance();
    }
    /**
     * question弹窗
     */
    question(content) {
        const confirmResult$ = this.confirmPromise(content);
        return from(confirmResult$);
    }
    /**
     * 带默认焦点的弹出提示
     * @param content 内容
     */
    question2(content) {
        const promise$ = new Promise((resolve, reject) => {
            const dialogRef = this.messagerService.question2(content, [
                {
                    text: this.languageService.no,
                    cls: 'btn btn-secondary',
                    handle: () => {
                        resolve(false);
                        dialogRef.close();
                    }
                },
                {
                    text: this.languageService.yes,
                    cls: 'btn btn-primary',
                    defaultFocus: true,
                    handle: () => {
                        resolve(true);
                        dialogRef.close();
                    }
                }
            ]);
        });
        return from(promise$);
    }
    /**
     * 弹出输入对话框
     * @param title 标题
     */
    prompt(title) {
        return this.messagerService.prompt(title);
    }
    /**
     * 使用Promise实例包装回调方法，以同步回调方法执行后结果
     * @param content
     */
    confirmPromise(content) {
        return new Promise((resolve, reject) => {
            this.messagerService.question(content, () => { resolve(true); }, () => { resolve(false); });
        });
    }
    /**
     * 确认弹框
     * @param content 弹出内容提示
     */
    confirm(content) {
        return this.messagerService.confirm(content);
    }
    /**
     * 消息弹框
     */
    info(content) {
        this.messagerService.info(content);
    }
    /**
     * 错误弹框
     */
    error(content) {
        this.messagerService.error(content);
    }
    /**
     * 警告弹框
     */
    warning(content) {
        this.messagerService.warning(content);
    }
    /**
     * 服务器端异常提示弹框
     */
    httpErrorInServer(httpError) {
        let error = httpError.error;
        if (typeof (error) === 'string') {
            try {
                error = JSON.parse(error);
            }
            catch (e) { }
        }
        if (!error || error.Level == null || error.Level == undefined) { // 未封装error或未设置异常等级
            this.messagerService.error(httpError.message);
            return;
        }
        const exceptionStrategy = ExceptionFactory.getInstance(this.messagerService, this.languageService).getExceptionHandleStrategy(error.Level);
        exceptionStrategy.handleException(error);
    }
    /**
     * 客户端异常提示弹框
     */
    httpErrorInClient(httpError) {
        if (!httpError) {
            return;
        }
        const message = httpError.error && httpError.error.error && httpError.error.error.message;
        const exceptionMsg = message || httpError.message && httpError.message.replace(/http:\/\/[a-zA-Z0-9.:]{1,}/, '');
        const options = {
            // title: this.languageService.clientNotifyTitle,
            // showMaxButton: true,
            showHeader: false,
            width: 400,
            height: 200,
            safeHtml: false
        };
        this.messagerService.show('error', exceptionMsg, options);
    }
    /**
     * 401异常处理
     * 为了独立弹出重新登录提示框，切换到运行框架http服务后可以删除
     * @param httpError 异常
     */
    http401Error(httpError) {
        const dialogShownKey = '401ErrorShownFlag';
        if (!httpError || window[dialogShownKey]) {
            return;
        }
        const i18n = {
            en: {
                title: 'Warning',
                sessionInvalid: 'Your login has expired, please login again.',
                ok: 'ok',
                cancel: 'cancel'
            },
            'zh-CHS': {
                title: '提示',
                sessionInvalid: '用户登录信息已失效，是否重新登录?',
                ok: '确认',
                cancel: '取消'
            }
        };
        this.curLanguage = this.curLanguage || 'zh-CHS';
        const messageOptions = {
            title: i18n[this.curLanguage].title,
            initialState: {
                okText: i18n[this.curLanguage].ok,
                okHandle: () => {
                    modalRef.close();
                    window[dialogShownKey] = false;
                    const eventResult = httpError && httpError.error || {};
                    const redirectUri = eventResult.redirect_uri || '/login.html';
                    window.location.href = `/logout.html#?logout-before-redirect=true&loginUri=${redirectUri}`;
                },
                cancelText: i18n[this.curLanguage].cancel,
                cancelHandle: () => {
                    modalRef.close();
                    window[dialogShownKey] = false;
                }
            },
            showHeader: true,
            width: 420,
            height: 180,
            fitContent: false
        };
        const modalRef = this.messagerService.show('question', i18n[this.curLanguage].sessionInvalid, messageOptions);
        window[dialogShownKey] = true;
        if (modalRef && modalRef.dialog && modalRef.dialog.instance.closed) {
            modalRef.dialog.instance.closed.subscribe(() => {
                window[dialogShownKey] = false;
            });
        }
    }
}
FormMessageService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FormMessageService.ctorParameters = () => [
    { type: MessagerService },
    { type: LanguageService, decorators: [{ type: Optional }] },
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
];
export { FormMessageService };
