// @farris/command-services/esm2015/lib/sidebar-service.js
import { Injectable, Optional } from '@angular/core';
import { of } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { FarrisSidebarService } from '@farris/ui-sidebar';
import { FrameContext } from '@farris/devkit';
import { FormMessageService } from './form-message.service';
import { LanguageService } from './languag.service';
/**
 * 侧边栏服务
 * @scope FrameComponent
 */
class SidebarService {
    constructor(frameContext, sidebarUIService, messageService, languageService) {
        this.frameContext = frameContext;
        this.sidebarUIService = sidebarUIService;
        this.messageService = messageService;
        this.languageService = languageService;
        if (!this.languageService) {
            this.languageService = LanguageService.getInstance();
        }
    }
    get repository() {
        return this.frameContext.repository;
    }
    openSidebar() {
        this.sidebarUIService.sendIsOpen(true);
    }
    closeSidebar() {
        this.sidebarUIService.sendIsOpen(false);
    }
    /**
     * 关闭前先取消变更
     */
    confirmBeforeClosingSidebar() {
        // 检查是否有变更，没有变更直接关闭
        const ifChangesExist = this.repository.entityManager.checkAllEntityChanges();
        if (!ifChangesExist) {
            return of(true);
        }
        // 确认是否关闭
        const confirm$ = this.messageService.question(this.languageService['exitWithoutSave']);
        const result$ = confirm$.pipe(switchMap((confirmResult) => {
            if (confirmResult === false) {
                return of(false);
            }
            else {
                return of(true);
            }
        }));
        return result$;
    }
    /**
     * 继续关闭侧边栏
     */
    continueClosingSidebar() {
        return of(true);
    }
    /**
     * 阻止侧边栏关闭
     */
    stopClosingSidebar() {
        return of(false);
    }
}
SidebarService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
SidebarService.ctorParameters = () => [
    { type: FrameContext },
    { type: FarrisSidebarService },
    { type: FormMessageService },
    { type: LanguageService, decorators: [{ type: Optional }] }
];
export { SidebarService };
