// @farris/command-services/esm2015/lib/user-settings.service.js
import { Injectable, Injector, Optional } from '@angular/core';
import { FrmI18nSettingService } from '@gsp-sys/rtf-common';
export class UserSettingsService {
    constructor(injector, frmI18nSettingService) {
        this.injector = injector;
        this.frmI18nSettingService = frmI18nSettingService;
    }
    /**
     * 用户时区字符串
     */
    get timeZone() {
        return this.timezone && this.timezone.id || null;
    }
    /**
     * 时区偏移
     */
    get timeZoneOffset() {
        const offset = this.timezone && this.timezone.baseOffset;
        if (offset !== null && offset !== undefined) {
            return offset;
        }
        else {
            return null;
        }
    }
    /**
     * 默认语言
     */
    get defaultLanguage() {
        return this.i18nSetting && this.i18nSetting.defaultLanguage || null;
    }
    /**
     * 日期格式化字符串
     * like: yyyy-MM-dd
     */
    get dateFormat() {
        return this.i18nSetting && this.i18nSetting.userDateFormat || null;
    }
    /**
     * 时间格式
     * like: HH:MM:ss
     */
    get timeFormat() {
        return this.i18nSetting && this.i18nSetting.userTimeFormat || null;
    }
    get dateTimeFormat() {
        const dateFormat = this.i18nSetting && this.i18nSetting.userDateFormat || 'yyyy-MM-dd';
        const timeFormat = this.i18nSetting && this.i18nSetting.userTimeFormat || 'HH:mm:ss';
        return `${dateFormat} ${timeFormat}`;
    }
    /**
     * 地址格式
     */
    get addressFormat() {
        return null;
    }
    /**
     * 数字格式
     */
    get numberFormat() {
        if (this.i18nSetting && this.i18nSetting.userNumberFormat) {
            const userNumberFormat = this.i18nSetting.userNumberFormat;
            const numberFormat = {
                negativeSign: userNumberFormat.negativeSign || '-',
                numberDecimalDigits: userNumberFormat.numberDecimalDigits || 2,
                numberDecimalSeparator: userNumberFormat.numberDecimalSeparator || '.',
                numberGroupSeparator: userNumberFormat.numberGroupSeparator || ','
            };
            return numberFormat;
        }
        return null;
    }
    /**
     * 区域代码
     */
    get regionCode() {
        return this.i18nSetting && this.i18nSetting.userRegionCode || null;
    }
    /**
     * 用户配置
     */
    get userSettings() {
        const userSettings = this.frmI18nSettingService && this.frmI18nSettingService.getSetting();
        return userSettings;
    }
    get timezone() {
        return this.i18nSetting && this.i18nSetting.timezone;
    }
    get i18nSetting() {
        return this.userSettings && this.userSettings.i18nSetting;
    }
}
UserSettingsService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
UserSettingsService.ctorParameters = () => [
    { type: Injector },
    { type: FrmI18nSettingService, decorators: [{ type: Optional }] }
];
