// @gsp-sys/rtf-common/esm2015/lib/services/frmEventService/framework-event.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { EventUtil, AppType } from '../../shared/frm-util';
import { FrmEventBus } from '../../eventbus/frm-evenbus';
import { Subject } from 'rxjs';
export class FrameworkEventService {
    /**
     * @param {?} feb
     */
    constructor(feb) {
        this.feb = feb;
        this.FarrisApps = EventUtil.FARRIS_APPS;
        this.FarrisAppClick = EventUtil.FARRIS_APP_CLICK;
        this.AllGspFuncs = EventUtil.ALL_GSP_FUNCS;
        this.BeforeFuncClose = EventUtil.BEFORE_FUNC_CLOSE;
        this.FuncClosed = EventUtil.FUNC_CLOSED;
        this.FuncSwitch = EventUtil.FUNC_SWITCH;
        this.AfterFrameLoad = EventUtil.AFTER_FRAME_LOAD;
        /** @type {?} */
        const eventBusEntity = EventUtil.getFrmEvent();
        if (!eventBusEntity) {
            EventUtil.setFrmEvent(this.feb || new FrmEventBus());
            this.eventInitial();
        }
    }
    /**
     * @param {?} token
     * @param {?} args
     * @return {?}
     */
    eventFire(token, args) {
        /** @type {?} */
        const feb = EventUtil.getFrmEvent();
        if (!feb || typeof feb.post !== 'function') {
            return;
        }
        feb.post(token, args);
    }
    /**
     * @param {?} token
     * @param {?} handler
     * @param {?=} options
     * @return {?}
     */
    eventListner(token, handler, options) {
        /** @type {?} */
        const feb = EventUtil.getFrmEvent();
        if (!feb || typeof feb.on !== 'function') {
            return;
        }
        // 菜单监听者必须要声明自己的身份，方便框架管理
        if (options && options.appType) {
            switch (options.appType.toLowerCase()) {
                case AppType.App:
                    /** @type {?} */
                    const appObkey = EventUtil.appObsToken(options.appId, options.appEntrance, options.tabId);
                    feb.onWithObkey(token, appObkey, handler);
                    return;
                case AppType.Menu:
                    /** @type {?} */
                    const funcObkey = EventUtil.funcObsToken(options.funcId, options.tabId);
                    feb.onWithObkey(token, funcObkey, handler);
                    return;
                default: break;
            }
        }
        feb.on(token, handler);
    }
    /**
     * @private
     * @return {?}
     */
    eventInitial() {
        /** @type {?} */
        const feb = EventUtil.getFrmEvent();
        if (!feb || typeof feb.post !== 'function') {
            return;
        }
        feb.post(this.FarrisApps, null);
        feb.post(this.FarrisAppClick, null);
        feb.post(this.AllGspFuncs, null);
        feb.initialPipe(this.FuncClosed, new Subject());
        feb.initialPipe(this.FuncSwitch, new Subject());
        feb.initialPipe(this.BeforeFuncClose, new Subject());
        feb.initialPipe(this.AfterFrameLoad, new Subject());
    }
}
FrameworkEventService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FrameworkEventService.ctorParameters = () => [
    { type: FrmEventBus }
];
if (false) {
    /** @type {?} */
    FrameworkEventService.prototype.FarrisApps;
    /** @type {?} */
    FrameworkEventService.prototype.FarrisAppClick;
    /** @type {?} */
    FrameworkEventService.prototype.AllGspFuncs;
    /** @type {?} */
    FrameworkEventService.prototype.BeforeFuncClose;
    /** @type {?} */
    FrameworkEventService.prototype.FuncClosed;
    /** @type {?} */
    FrameworkEventService.prototype.FuncSwitch;
    /** @type {?} */
    FrameworkEventService.prototype.AfterFrameLoad;
    /**
     * @type {?}
     * @private
     */
    FrameworkEventService.prototype.feb;
}
