// @gsp-sys/rtf-common/esm5/lib/services/funcsservice/funcs.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { FrameworkVariableService } from '../frmVariableService/framework-variable.service';
var FuncsService = /** @class */ (function () {
    function FuncsService(frmVariableService) {
        this.frmVariableService = frmVariableService;
        this.allgspfuncs = 'allgspfuncs';
        this.allfuncinvoks = 'allfuncinvoks';
        if (!this.frmVariableService) {
            this.frmVariableService = new FrameworkVariableService();
        }
    }
    /**
     * 缓存所有功能
     */
    /**
     * 缓存所有功能
     * @param {?} funcs
     * @return {?}
     */
    FuncsService.prototype.cacheAllFuncs = /**
     * 缓存所有功能
     * @param {?} funcs
     * @return {?}
     */
    function (funcs) {
        this.frmVariableService.frameworkVariableReg(this.allgspfuncs, funcs);
    };
    /**
     * 获取所有缓存的功能
     */
    /**
     * 获取所有缓存的功能
     * @return {?}
     */
    FuncsService.prototype.getAllCachedFuncs = /**
     * 获取所有缓存的功能
     * @return {?}
     */
    function () {
        /** @type {?} */
        var funcsCache = this.frmVariableService.getFrmVariableByKey(this.allgspfuncs);
        return (/** @type {?} */ (funcsCache)) || [];
    };
    /**
     * 获取某功能
     */
    /**
     * 获取某功能
     * @param {?} funcId
     * @return {?}
     */
    FuncsService.prototype.getCachedFunc = /**
     * 获取某功能
     * @param {?} funcId
     * @return {?}
     */
    function (funcId) {
        /** @type {?} */
        var allfuncs = this.getAllCachedFuncs();
        if (allfuncs) {
            return allfuncs.find((/**
             * @param {?} f
             * @return {?}
             */
            function (f) { return f.id === funcId; }));
        }
    };
    /**
     * @param {?} invoks
     * @return {?}
     */
    FuncsService.prototype.cacheAllFuncInvoks = /**
     * @param {?} invoks
     * @return {?}
     */
    function (invoks) {
        this.frmVariableService.frameworkVariableReg(this.allfuncinvoks, invoks);
    };
    /**
     * @return {?}
     */
    FuncsService.prototype.getAllCachedFuncInvoks = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var invokCache = this.frmVariableService.getFrmVariableByKey(this.allfuncinvoks);
        return (/** @type {?} */ (invokCache)) || [];
    };
    FuncsService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FuncsService.ctorParameters = function () { return [
        { type: FrameworkVariableService, decorators: [{ type: Optional }] }
    ]; };
    return FuncsService;
}());
export { FuncsService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FuncsService.prototype.allgspfuncs;
    /**
     * @type {?}
     * @private
     */
    FuncsService.prototype.allfuncinvoks;
    /**
     * @type {?}
     * @private
     */
    FuncsService.prototype.frmVariableService;
}
