// @gspwidget/widget-devkit/esm5/lib/data.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, EventEmitter } from '@angular/core';
/**
 * @record
 */
export function LoadDataOptions() { }
if (false) {
    /** @type {?|undefined} */
    LoadDataOptions.prototype.param;
}
/**
 * 部件系统、数据服务内部使用
 * @record
 */
export function LoadDataPayload() { }
/**
 * 数据加载不应该让此 service 直接发出，因为外部框架并不随时都支持取数，
 * 比如新增部件后的部件配置，此时没有配置过数据，便会服务端报错。
 * 应该交由元部件容器加载数据。
 */
var WidgetDataService = /** @class */ (function () {
    function WidgetDataService() {
        /**
         * loadData() / reloadData() 时触发，外部元部件容器接收、加载数据，并通过 setData 设置回数据
         */
        this.onLoadData = new EventEmitter();
        this.onGetData = new EventEmitter();
        this.onGetDataWithDims = new EventEmitter();
    }
    /**
     * 元部件调用此方法获取数据。示例：
     * ```ts
     * this.dataService.loadData().subscribe(data => {
     *  this.data = data // 使用数据
     * })
     * ```
     */
    /**
     * 元部件调用此方法获取数据。示例：
     * ```ts
     * this.dataService.loadData().subscribe(data => {
     *  this.data = data // 使用数据
     * })
     * ```
     * @param {?=} options
     * @return {?}
     */
    WidgetDataService.prototype.loadData = /**
     * 元部件调用此方法获取数据。示例：
     * ```ts
     * this.dataService.loadData().subscribe(data => {
     *  this.data = data // 使用数据
     * })
     * ```
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        this.onLoadData.emit(options);
        return this.onGetData;
    };
    /**
     * 获取全数据信息，包含 dimensions
     */
    /**
     * 获取全数据信息，包含 dimensions
     * @param {?=} options
     * @return {?}
     */
    WidgetDataService.prototype.loadDataWithDims = /**
     * 获取全数据信息，包含 dimensions
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        this.onLoadData.emit(options);
        return this.onGetDataWithDims;
    };
    /**
     * 元部件调用此方法以重新获取数据，数据加载完成后再次调用在 loadData() 注册的事件处理器。示例：
     * ```ts
     * this.dataService.refreshData()
     * ```
     */
    /**
     * 元部件调用此方法以重新获取数据，数据加载完成后再次调用在 loadData() 注册的事件处理器。示例：
     * ```ts
     * this.dataService.refreshData()
     * ```
     * @param {?=} param
     * @return {?}
     */
    WidgetDataService.prototype.reloadData = /**
     * 元部件调用此方法以重新获取数据，数据加载完成后再次调用在 loadData() 注册的事件处理器。示例：
     * ```ts
     * this.dataService.refreshData()
     * ```
     * @param {?=} param
     * @return {?}
     */
    function (param) {
        this.onLoadData.emit(param);
    };
    /**
     * 外部通过此方法给元部件设置数据，元部件无需调用
     * @param dataWithDims 要设置的数据
     */
    /**
     * 外部通过此方法给元部件设置数据，元部件无需调用
     * @param {?} dataWithDims 要设置的数据
     * @return {?}
     */
    WidgetDataService.prototype.setData = /**
     * 外部通过此方法给元部件设置数据，元部件无需调用
     * @param {?} dataWithDims 要设置的数据
     * @return {?}
     */
    function (dataWithDims) {
        this.onGetData.emit(dataWithDims.tableData);
        this.onGetDataWithDims.emit(dataWithDims);
    };
    WidgetDataService.decorators = [
        { type: Injectable }
    ];
    return WidgetDataService;
}());
export { WidgetDataService };
if (false) {
    /**
     * loadData() / reloadData() 时触发，外部元部件容器接收、加载数据，并通过 setData 设置回数据
     * @type {?}
     */
    WidgetDataService.prototype.onLoadData;
    /**
     * @type {?}
     * @private
     */
    WidgetDataService.prototype.onGetData;
    /**
     * @type {?}
     * @private
     */
    WidgetDataService.prototype.onGetDataWithDims;
}
