// @gspwidget/widget-devkit/esm5/lib/WidgetBase.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// import { datalessJumpOption, DatalessJumpConfig, JumpTypes } from './presets';
import { Property } from './widget-registry/widget-property.decorator';
/**
 * @abstract
 */
var WidgetBase = /** @class */ (function () {
    /**
     * 因为是父类构造函数所以无法依赖注入, 也没法传入参数;
     * compInstance.field = xxx 的方式又发生在构造函数后而不太好,
     * 所以考虑此父类不提供成员属性, 而通过依赖注入, 在子组件自取.
     */
    function WidgetBase() {
        this.title = 'untitled';
        /**
         * @deprecated 是否使用外部数据源。是的话，wc 就拿它的 portlet id 请求数据去。应改为 dataEngine 写法。
         */
        this.useDataEngine = false;
        if (this.widgetName == null)
            console.error("\u5143\u90E8\u4EF6 " + this.constructor.name + " \u672A\u6DFB\u52A0 @Widget({ name: <name>})");
    }
    /**
     * 元部件容器大小变化事件，参数 width 为容器宽度，单位 px。
     * 可重写此方法以处理复杂的响应式样式。
     */
    /**
     * 元部件容器大小变化事件，参数 width 为容器宽度，单位 px。
     * 可重写此方法以处理复杂的响应式样式。
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    WidgetBase.prototype.onResized = /**
     * 元部件容器大小变化事件，参数 width 为容器宽度，单位 px。
     * 可重写此方法以处理复杂的响应式样式。
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    function (width, height) { };
    /**
     * @param {?} propName
     * @param {?} value
     * @return {?}
     */
    WidgetBase.prototype.onPropertyChange = /**
     * @param {?} propName
     * @param {?} value
     * @return {?}
     */
    function (propName, value) { };
    /**
     * @deprecated
     * 请用 this.dataService.loadData() 代替
     * @param data 获得的数据
     */
    /**
     * @deprecated
     * 请用 this.dataService.loadData() 代替
     * @param {?} data 获得的数据
     * @return {?}
     */
    WidgetBase.prototype.onGetData = /**
     * @deprecated
     * 请用 this.dataService.loadData() 代替
     * @param {?} data 获得的数据
     * @return {?}
     */
    function (data) { };
    tslib_1.__decorate([
        Property({
            category: '公共属性',
            displayName: '部件标题'
        }),
        tslib_1.__metadata("design:type", Object)
    ], WidgetBase.prototype, "title", void 0);
    return WidgetBase;
}());
export { WidgetBase };
if (false) {
    /** @type {?} */
    WidgetBase.prototype.title;
    /**
     * 编译时将 \@Widget 中的 name 赋值给 widgetName, 构造器中会检查是否有值，没有就报错
     * @type {?}
     */
    WidgetBase.prototype.widgetName;
    /**
     * @deprecated 当前部件 id
     * @type {?}
     */
    WidgetBase.prototype.portletId;
    /**
     * @deprecated 当前部件实例 id
     * @type {?}
     */
    WidgetBase.prototype.id;
    /**
     * @deprecated 是否使用外部数据源。是的话，wc 就拿它的 portlet id 请求数据去。应改为 dataEngine 写法。
     * @type {?}
     */
    WidgetBase.prototype.useDataEngine;
    /**
     * @deprecated
     * assets 文件路径必需加上这个 base 前缀。注意在构造函数或之前使用时其值会为 undefined，
     * 因为实例化完成后，widget-container 才会给它赋正确的值。
     * @type {?}
     */
    WidgetBase.prototype.assetsBaseUrl;
}
