// @gspwidget/widget-devkit/esm5/lib/widget-devkit.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { JumpTypes } from './preset/preset.jump';
import { CHART_TABLE_FUNC_ID } from './util';
/**
 * @deprecated
 * 为了 widget-devkit 的兼容而暂时保留。不准再用了。
 * 元部件应该直接用 \@gspwidget/util。
 * @type {?}
 */
export var DASHBOARD_API = 'DashboardApi';
var WidgetDevkitService = /** @class */ (function () {
    function WidgetDevkitService(modalService) {
        this.modalService = modalService;
    }
    /**
     * @param {?} content
     * @param {?=} config
     * @return {?}
     */
    WidgetDevkitService.prototype.openModal = /**
     * @param {?} content
     * @param {?=} config
     * @return {?}
     */
    function (content, config) {
        return this.modalService.show(content, config);
    };
    /**
     * @param {?} funcId
     * @param {?=} queryParamsArray
     * @return {?}
     */
    WidgetDevkitService.prototype.openMenu = /**
     * @param {?} funcId
     * @param {?=} queryParamsArray
     * @return {?}
     */
    function (funcId, queryParamsArray) {
        window[DASHBOARD_API].openMenu(funcId, queryParamsArray);
    };
    /**
     * @param {?} appId
     * @return {?}
     */
    WidgetDevkitService.prototype.openApp = /**
     * @param {?} appId
     * @return {?}
     */
    function (appId) {
        window[DASHBOARD_API].openApp(appId);
    };
    /**
     * options 参考 https://ng.ant.design/components/message/zh
     */
    /**
     * options 参考 https://ng.ant.design/components/message/zh
     * @param {?} content
     * @param {?=} options
     * @return {?}
     */
    WidgetDevkitService.prototype.openInfoMsg = /**
     * options 参考 https://ng.ant.design/components/message/zh
     * @param {?} content
     * @param {?=} options
     * @return {?}
     */
    function (content, options) {
        window[DASHBOARD_API].messageSvc.info(content, options);
    };
    /**
     * options 参考 https://ng.ant.design/components/message/zh
     */
    /**
     * options 参考 https://ng.ant.design/components/message/zh
     * @param {?} content
     * @param {?=} options
     * @return {?}
     */
    WidgetDevkitService.prototype.openSuccessMsg = /**
     * options 参考 https://ng.ant.design/components/message/zh
     * @param {?} content
     * @param {?=} options
     * @return {?}
     */
    function (content, options) {
        window[DASHBOARD_API].messageSvc.success(content, options);
    };
    /**
     * options 参考 https://ng.ant.design/components/message/zh
     */
    /**
     * options 参考 https://ng.ant.design/components/message/zh
     * @param {?} content
     * @param {?=} options
     * @return {?}
     */
    WidgetDevkitService.prototype.openWarningMsg = /**
     * options 参考 https://ng.ant.design/components/message/zh
     * @param {?} content
     * @param {?=} options
     * @return {?}
     */
    function (content, options) {
        window[DASHBOARD_API].messageSvc.warning(content, options);
    };
    /**
     * options 参考 https://ng.ant.design/components/message/zh
     */
    /**
     * options 参考 https://ng.ant.design/components/message/zh
     * @param {?} content
     * @param {?=} options
     * @return {?}
     */
    WidgetDevkitService.prototype.openErrorMsg = /**
     * options 参考 https://ng.ant.design/components/message/zh
     * @param {?} content
     * @param {?=} options
     * @return {?}
     */
    function (content, options) {
        window[DASHBOARD_API].messageSvc.error(content, options);
    };
    /**
     * @param {?} dataOption
     * @return {?}
     */
    WidgetDevkitService.prototype.getData = /**
     * @param {?} dataOption
     * @return {?}
     */
    function (dataOption) {
        return window[DASHBOARD_API].storageSvc.getData(dataOption);
    };
    /**
     * @param {?=} reload
     * @return {?}
     */
    WidgetDevkitService.prototype.getUserInfo = /**
     * @param {?=} reload
     * @return {?}
     */
    function (reload) {
        if (reload === void 0) { reload = false; }
        return window[DASHBOARD_API].storageSvc.getUserInfo(reload);
    };
    /**
     * @param {?=} reload
     * @return {?}
     */
    WidgetDevkitService.prototype.getUserAvatar = /**
     * @param {?=} reload
     * @return {?}
     */
    function (reload) {
        if (reload === void 0) { reload = false; }
        return window[DASHBOARD_API].storageSvc.getUserAvatar(reload);
    };
    /**
     * @deprecated 应用另外两个 jump 方法
     * 联查跳转，直接把配置属性的值传进来就行。配成“无”就不跳转。
     * @param jumpProperty 跳转的配置属性的值
     * @param dataItem 当前跳转对应的数据行。如果是全局跳转，就传入第一行？
     */
    /**
     * @deprecated 应用另外两个 jump 方法
     * 联查跳转，直接把配置属性的值传进来就行。配成“无”就不跳转。
     * @param {?} jumpProperty 跳转的配置属性的值
     * @param {?=} dataItem 当前跳转对应的数据行。如果是全局跳转，就传入第一行？
     * @return {?}
     */
    WidgetDevkitService.prototype.jump = /**
     * @deprecated 应用另外两个 jump 方法
     * 联查跳转，直接把配置属性的值传进来就行。配成“无”就不跳转。
     * @param {?} jumpProperty 跳转的配置属性的值
     * @param {?=} dataItem 当前跳转对应的数据行。如果是全局跳转，就传入第一行？
     * @return {?}
     */
    function (jumpProperty, dataItem) {
        var type = jumpProperty.type, funcId = jumpProperty.funcId, params = jumpProperty.params;
        /** @type {?} */
        var queryParamsArray = Array.isArray(params)
            ? params.map((/**
             * @param {?} param
             * @return {?}
             */
            function (param) { return [param.key, (param.valueType === 'fixedValue'
                    ? param.fixedValue
                    : (dataItem
                        ? dataItem[param.fieldName.name]
                        : undefined))]; }))
            : undefined;
        switch (type) {
            case JumpTypes.None:
                return;
            case JumpTypes.Func:
                this.openMenu(funcId, queryParamsArray);
                break;
            case JumpTypes.ChartTable:
                this.openMenu(CHART_TABLE_FUNC_ID, queryParamsArray);
                break;
        }
    };
    /**
     * @param {?} dataJumpConfig
     * @param {?} dataItem
     * @return {?}
     */
    WidgetDevkitService.prototype.jumpWithData = /**
     * @param {?} dataJumpConfig
     * @param {?} dataItem
     * @return {?}
     */
    function (dataJumpConfig, dataItem) {
        var jumpType = dataJumpConfig.jumpType, funcIdSource = dataJumpConfig.funcIdSource, funcIdDataField = dataJumpConfig.funcIdDataField, funcId = dataJumpConfig.funcId, funcParams = dataJumpConfig.funcParams, url = dataJumpConfig.url, urlParams = dataJumpConfig.urlParams, queryParams = dataJumpConfig.queryParams;
        /** @type {?} */
        var funcParamsArray = Array.isArray(funcParams)
            ? funcParams.map((/**
             * @param {?} funcParam
             * @return {?}
             */
            function (funcParam) { return [funcParam.key, (funcParam.valueType === 'fixedValue'
                    ? funcParam.fixedValue
                    : (dataItem
                        ? dataItem[funcParam.fieldName.name]
                        : undefined))]; }))
            : undefined;
        switch (jumpType) {
            case JumpTypes.None:
                return;
            case JumpTypes.Func:
                /** @type {?} */
                var resultFuncId = (!funcIdSource || funcIdSource === 'fixed')
                    ? funcId
                    : dataItem[funcIdDataField.name];
                if (resultFuncId)
                    this.openMenu(resultFuncId, funcParamsArray);
                break;
            case JumpTypes.ChartTable:
                this.openMenu(CHART_TABLE_FUNC_ID, funcParamsArray);
                break;
            case JumpTypes.NewBrowserTab:
                // 拼 url
                /** @type {?} */
                var finalUrl = url
                // + urlParams
                ;
                // + urlParams
                if (urlParams && urlParams.length > 0) {
                    /** @type {?} */
                    var urlParamPart = urlParams.map((/**
                     * @param {?} urlParam
                     * @return {?}
                     */
                    function (urlParam) {
                        var valueType = urlParam.valueType, fixedValue = urlParam.fixedValue, fieldName = urlParam.fieldName;
                        return valueType === 'fixedValue' ? fixedValue : dataItem[fieldName.name];
                    })).join('/');
                    finalUrl += url.endsWith('/') ? '' : '/';
                    finalUrl += urlParamPart;
                }
                // + queryParams
                if (queryParams && queryParams.length > 0) {
                    /** @type {?} */
                    var queryParamPart = queryParams.map((/**
                     * @param {?} queryParam
                     * @return {?}
                     */
                    function (queryParam) {
                        var key = queryParam.key, valueType = queryParam.valueType, fixedValue = queryParam.fixedValue, fieldName = queryParam.fieldName;
                        return key + "=" + (valueType === 'fixedValue' ? fixedValue : dataItem[fieldName.name]);
                    })).join('&');
                    finalUrl += '?' + queryParamPart;
                }
                // 打开
                window.open(finalUrl, '_blank');
        }
    };
    /**
     * @param {?} datalessJumpConfig
     * @return {?}
     */
    WidgetDevkitService.prototype.jumpWithoutData = /**
     * @param {?} datalessJumpConfig
     * @return {?}
     */
    function (datalessJumpConfig) {
        var jumpType = datalessJumpConfig.jumpType, funcId = datalessJumpConfig.funcId, funcParams = datalessJumpConfig.funcParams, url = datalessJumpConfig.url;
        /** @type {?} */
        var funcParamsArray = Array.isArray(funcParams)
            ? funcParams.map((/**
             * @param {?} funcParam
             * @return {?}
             */
            function (funcParam) { return [funcParam.key, funcParam.value]; }))
            : undefined;
        switch (jumpType) {
            case JumpTypes.None:
                return;
            case JumpTypes.Func:
                this.openMenu(funcId, funcParamsArray);
                break;
            case JumpTypes.ChartTable:
                this.openMenu(CHART_TABLE_FUNC_ID, funcParamsArray);
                break;
            case JumpTypes.NewBrowserTab:
                window.open(url, '_blank');
        }
    };
    WidgetDevkitService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    WidgetDevkitService.ctorParameters = function () { return [
        { type: BsModalService }
    ]; };
    return WidgetDevkitService;
}());
export { WidgetDevkitService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    WidgetDevkitService.prototype.modalService;
}
