// @farris/ide-devkit/esm2015/lib/ideframework/gsp/event-kit/composite-disposable.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Disposable } from './disposable';
// Essential: An object that aggregates multiple {Disposable} instances together
// into a single disposable, so they can all be disposed as a group.
//
// These are very useful when subscribing to multiple events.
//
// ## Examples
//
// ```coffee
// {CompositeDisposable} = require 'atom'
//
// class Something
//   constructor: ->
//     @disposables = new CompositeDisposable
//     editor = atom.workspace.getActiveTextEditor()
//     @disposables.add editor.onDidChange ->
//     @disposables.add editor.onDidChangePath ->
//
//   destroy: ->
//     @disposables.dispose()
// ```
export class CompositeDisposable {
    //////
    /// Section: Construction and Destruction
    //////
    // Public: Construct an instance, optionally with one or more disposables
    /**
     * @param {...?} disposables
     */
    constructor(...disposables) {
        this.disposed = false;
        this.disposables = new Set();
        this.add(...disposables);
    }
    // Public: Dispose all disposables added to this composite disposable.
    //
    // If this object has already been disposed, this method has no effect.
    /**
     * @return {?}
     */
    dispose() {
        if (!this.disposed) {
            this.disposed = true;
            this.disposables.forEach((disposable) => disposable.dispose());
            this.disposables = null;
        }
    }
    //////
    /// Section: Managing Disposables
    //////
    // Public: Add disposables to be disposed when the composite is disposed.
    //
    // If this object has already been disposed, this method has no effect.
    //
    // * `...disposables` {Disposable} instances or any objects with `.dispose()`
    //   methods.
    /**
     * @param {...?} disposables
     * @return {?}
     */
    add(...disposables) {
        if (!this.disposed) {
            for (const disposable of disposables) {
                if (this.assertDisposable(disposable)) {
                    this.disposables.add(disposable);
                }
            }
        }
    }
    // Public: Remove a previously added disposable.
    //
    // * `disposable` {Disposable} instance or any object with a `.dispose()`
    //   method.
    /**
     * @param {?} disposable
     * @return {?}
     */
    remove(disposable) {
        if (!this.disposed) {
            this.disposables.delete(disposable);
        }
    }
    // Public: Alias to {CompositeDisposable::remove}
    /**
     * @param {?} disposable
     * @return {?}
     */
    delete(disposable) {
        this.remove(disposable);
    }
    // Public: Clear all disposables. They will not be disposed by the next call
    // to dispose.
    /**
     * @return {?}
     */
    clear() {
        if (!this.disposed) {
            this.disposables.clear();
        }
    }
    /**
     * @param {?} disposable
     * @return {?}
     */
    assertDisposable(disposable) {
        if (!Disposable.isDisposable(disposable)) {
            throw new TypeError('Arguments to CompositeDisposable.add must have a .dispose() method');
        }
        return true;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    CompositeDisposable.prototype.disposed;
    /**
     * @type {?}
     * @private
     */
    CompositeDisposable.prototype.disposables;
}
