// @farris/dynamic-control-group/esm2015/lib/components/input/input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from "@angular/core";
import { FormGroup } from '@angular/forms';
export class InputComponent {
    constructor() { }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.className = this.control.className ? this.control.className : 'col-12 col-md-6 col-xl-3 col-el-2';
    }
}
InputComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-input",
                template: `
      <div class="farris-group-wrap" [formGroup]="form">
        <div class="form-group farris-form-group">
          <label class="col-form-label">
            <span *ngIf="0" class="farris-label-info text-danger">*</span>
            <span class="farris-label-text">{{control.name}}</span>
          </label>
          <div class="farris-input-wrap">
            <input-group [formControlName]="control.labelCode" [readonly]="control.control.readonly"></input-group>
          </div>
        </div>
      </div>
  `
            }] }
];
/** @nocollapse */
InputComponent.ctorParameters = () => [];
InputComponent.propDecorators = {
    className: [{ type: HostBinding, args: ['class',] }],
    control: [{ type: Input }],
    form: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    InputComponent.prototype.className;
    /** @type {?} */
    InputComponent.prototype.control;
    /** @type {?} */
    InputComponent.prototype.form;
}
