// @farris/ui-datagrid/esm2015/lib/services/virtualized-loader.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class VirtualizedLoaderService {
    /**
     * @return {?}
     */
    getTableHeight() {
        return this.state.height;
    }
    /**
     * @return {?}
     */
    getTableHeaderHeight() {
        return this.state.headerHeight;
    }
    /**
     * @return {?}
     */
    getTableBodyHeight() {
        return this.getTableHeight() - this.getTableHeaderHeight();
    }
    /**
     * @return {?}
     */
    getTableWidth() {
        return this.state.width;
    }
    /**
     * @return {?}
     */
    getRowHeight() {
        return this.state.rowHeight;
    }
    /**
     * @return {?}
     */
    displayRowsCount() {
        return Math.floor(this.getTableHeight() / this.getRowHeight());
    }
    /**
     * @return {?}
     */
    getTotal() {
        /** @type {?} */
        let total = this.state.total;
        /** @type {?} */
        const data = this.state.data || [];
        /** @type {?} */
        const pagination = this.state.pagination;
        /** @type {?} */
        const count = data.length;
        if (pagination) {
            total = count;
        }
        return total;
    }
    /**
     * @return {?}
     */
    reload() {
        /** @type {?} */
        const rowHeight = this.getRowHeight();
        /** @type {?} */
        const scroTop = this.state.virtual.scrollTop;
        /** @type {?} */
        const res = this.getRows(scroTop);
        /** @type {?} */
        const total = this.getTotal();
        res.topHideHeight = scroTop;
        if (res.bottomHideHeight !== 0) {
            res.bottomHideHeight = total * rowHeight - res.virtualRows.length * rowHeight - res.topHideHeight;
        }
        return res;
    }
    /**
     * @param {?} scrollTop
     * @return {?}
     */
    getRows(scrollTop) {
        /** @type {?} */
        const minTop = Math.abs(scrollTop);
        /** @type {?} */
        const rowHeight = this.getRowHeight();
        /** @type {?} */
        const maxTop = minTop + this.getTableHeight();
        /** @type {?} */
        let top = (!this.state.virtualizedAsyncLoad) ? 0 : this.state.virtual.rowIndex * rowHeight;
        /** @type {?} */
        const rows = [];
        /** @type {?} */
        let topHideHeight = 0;
        /** @type {?} */
        let bottomHideHeight = 0;
        /** @type {?} */
        const data = this.state.data;
        /** @type {?} */
        const total = this.getTotal();
        for (const n of data) {
            top += rowHeight;
            if (top + rowHeight < minTop) {
                topHideHeight += rowHeight;
                continue;
            }
            else {
                if (top > maxTop) {
                    bottomHideHeight += rowHeight;
                    continue;
                }
            }
            rows.push(n);
        }
        if (this.state.virtualizedAsyncLoad) {
            topHideHeight = this.state.virtual.rowIndex * rowHeight + topHideHeight;
            bottomHideHeight = total * rowHeight - rows.length * rowHeight - topHideHeight;
        }
        /** @type {?} */
        let startIndex = this.state.virtual.rowIndex;
        if (data && data.length && rows && rows.length) {
            /** @type {?} */
            const rid = rows[0][this.state.idField];
            startIndex = this.state.virtual.rowIndex + data.findIndex((/**
             * @param {?} r
             * @return {?}
             */
            r => r[this.state.idField] === rid));
        }
        return {
            startIndex,
            virtualRows: rows,
            topHideHeight,
            bottomHideHeight
        };
    }
    /**
     * @param {?} scrollTop
     * @param {?} itemsCount
     * @param {?} firstRowIndex
     * @return {?}
     */
    getRowsCount(scrollTop, itemsCount, firstRowIndex) {
        /** @type {?} */
        const rowHeight = this.getRowHeight();
        /** @type {?} */
        const total = this.getTotal();
        /** @type {?} */
        const maxTop = scrollTop + this.getTableBodyHeight();
        /** @type {?} */
        let top = firstRowIndex * rowHeight;
        /** @type {?} */
        let rowsLength = 0;
        /** @type {?} */
        let topHideHeight = 0;
        /** @type {?} */
        let bottomHideHeight = 0;
        for (let i = 0; i < itemsCount; i++) {
            top += rowHeight;
            if (top + rowHeight < scrollTop) {
                topHideHeight += rowHeight;
                continue;
            }
            else {
                if (top > maxTop) {
                    continue;
                }
            }
            rowsLength++;
        }
        bottomHideHeight = total * rowHeight - rowsLength * rowHeight - topHideHeight;
        return { rowsLength, top: topHideHeight, bottom: bottomHideHeight };
    }
}
if (false) {
    /** @type {?} */
    VirtualizedLoaderService.prototype.state;
}
