// @farris/ui-html-editor/esm5/lib/html-editor-defaults.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var defaultModules = {
    toolbar: [
        // ['bold', 'italic', 'underline', 'strike'], // toggled buttons
        ['bold', 'italic', 'code-block', { header: 1 }, { header: 2 }, { list: 'ordered' }, { list: 'bullet' }, 'image']
        // ['blockquote', 'code-block'],
        // [{ header: 1 }, { header: 2 }], // custom button values
        // [{ list: 'ordered' }, { list: 'bullet' }],
        // [{ script: 'sub' }, { script: 'super' }], // superscript/subscript
        // [{ indent: '-1' }, { indent: '+1' }], // outdent/indent
        // [{ direction: 'rtl' }], // text direction
        // [{ size: ['small', false, 'large', 'huge'] }], // custom dropdown
        // [{ header: [1, 2, 3, 4, 5, 6, false] }],
        // [{ color: [] }, { background: [] }], // dropdown with defaults from theme
        // [
        //     {
        //         font: [
        //             'Microsoft-YaHei',
        //             'SimSun',
        //             'SimHei',
        //             'KaiTi',
        //             'FangSong',
        //             'Arial',
        //             'Times',
        //             'sans-serif'
        //         ]
        //     }
        // ],
        // [{ align: [] }],
        // ['clean'], // remove formatting button
        // ['link', 'image', 'video'] // link and image, video,
    ],
    imageDrop: true,
    imageResize: true,
    tooltip: true
};
/** @type {?} */
export var toolbarOptions = {
    fontBar: [
        {
            font: ['Microsoft-YaHei', 'SimSun', 'SimHei', 'KaiTi', 'FangSong', 'Arial', 'sans-serif']
        }
    ],
    headerBar: [{ header: [1, 2, 3, 4, 5, 6, false] }],
    sizeBar: [{ size: ['small', false, 'large', 'huge'] }],
    fontStyleBar: ['bold', 'italic', 'underline', 'strike', 'link'],
    indentBar: [{ indent: '-1' }, { indent: '+1' }],
    listBar: [{ list: 'ordered' }, { list: 'bullet' }],
    BlockBar: ['code-block', 'blockquote'],
    superAndSubscriptBar: [{ script: 'sub' }, { script: 'super' }],
    colorBar: [{ color: [] }, { background: [] }],
    mediaBar: ['image', 'video'],
    alignBar: [{ align: [] }],
    imageBar: ['image'],
    videoBar: ['video'],
    cleanBar: ['clean']
};
