// @gsp-wf/wf-process-editor/esm5/lib/views/anonymous-set/anonymous-set.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, EventEmitter, Output, Input, ViewChild, TemplateRef, Inject, Injector, Optional, ChangeDetectorRef, LOCALE_ID } from '@angular/core';
import { AnonymousSet } from './anonymous-set.entity';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../i18n/pipe/wf-process-editor-locale.pipe';
var AnonymousSetComponent = /** @class */ (function () {
    function AnonymousSetComponent(injector, cdr, localeId, resources, localePipe) {
        this.injector = injector;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.cls = 'farris-main-area flex-column';
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.anonymousSet = new AnonymousSet();
        this.modalConfig = {
            title: this.localePipe.transform('component.anonymousSet.anonymousOpt'),
            width: 450,
            height: 360,
            showButtons: true
        };
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
    }
    Object.defineProperty(AnonymousSetComponent.prototype, "value", {
        set: /**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            this.anonymousSet.isAnonymity = data.isAnonymity || false;
            this.anonymousSet.anonymous = data.anonymous || this.localePipe.transform('component.anonymousSet.approver');
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    AnonymousSetComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    AnonymousSetComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var object = { value: this.anonymousSet, parameters: null };
        this.submitModal.emit(object);
    };
    /**
     * @return {?}
     */
    AnonymousSetComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeModal.emit();
    };
    AnonymousSetComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-anonymous-set',
                    template: "<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <div class=\"f-section-header pt-2 mx-2\">\r\n        <div class=\"f-title\">\r\n          <h4 class=\"f-title-text\">{{'component.anonymousSet.anonymousSet'|wfLocale}}</h4>\r\n        </div>\r\n      </div>\r\n      <div class=\"f-utils-fill-flex-column h-100\">\r\n        <div class=\"farris-form form-inline farris-form-inline\">\r\n          <div class=\"col-12 farris-group-auto my-3 pr-5\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{'component.anonymousSet.isAnonymity'|wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap ml-1\">\r\n                  <farris-switch [(ngModel)]=\"anonymousSet.isAnonymity\" name=\"isAnonymity\"></farris-switch>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 farris-group-auto mt-4 pr-5\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{'component.anonymousSet.anonymous'|wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap ml-1\">\r\n                  <textarea class=\"form-control\" type=\"text\" value=\"anonymous\" [(ngModel)]=\"anonymousSet.anonymous\" name=\"anonymous\"\r\n                  style=\"height: 50px;\" [readOnly]=\"!anonymousSet.isAnonymity\"></textarea>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </section>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm mx-2\" (click)=\"clickCancel()\">{{'component.anonymousSet.clickCancel'|wfLocale}}</button>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">{{'component.anonymousSet.clickOK'|wfLocale}}</button>\r\n</ng-template>\r\n",
                    providers: [
                        WfLocalePipe,
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    AnonymousSetComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
        { type: WfLocalePipe, decorators: [{ type: Optional }] }
    ]; };
    AnonymousSetComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        closeModal: [{ type: Output }],
        submitModal: [{ type: Output }],
        value: [{ type: Input }],
        modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }]
    };
    return AnonymousSetComponent;
}());
export { AnonymousSetComponent };
if (false) {
    /** @type {?} */
    AnonymousSetComponent.prototype.cls;
    /** @type {?} */
    AnonymousSetComponent.prototype.closeModal;
    /** @type {?} */
    AnonymousSetComponent.prototype.submitModal;
    /** @type {?} */
    AnonymousSetComponent.prototype.modalFooter;
    /** @type {?} */
    AnonymousSetComponent.prototype.anonymousSet;
    /** @type {?} */
    AnonymousSetComponent.prototype.modalConfig;
    /**
     * @type {?}
     * @private
     */
    AnonymousSetComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    AnonymousSetComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    AnonymousSetComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    AnonymousSetComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    AnonymousSetComponent.prototype.localePipe;
}
var AnonymousSetConverter = /** @class */ (function () {
    function AnonymousSetConverter(localePipe) {
        this.localePipe = localePipe;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    AnonymousSetConverter.prototype.convertTo = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data && data.isAnonymity) {
            return this.localePipe.transform("component.anonymousSet.set");
        }
        else {
            return this.localePipe.transform("component.anonymousSet.notSet");
        }
    };
    return AnonymousSetConverter;
}());
export { AnonymousSetConverter };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AnonymousSetConverter.prototype.localePipe;
}
