// @gsp-wf/wf-process-editor/esm5/lib/views/execute-forms/add-execute-form/form-format.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { map } from 'rxjs/operators';
var FormFormatService = /** @class */ (function () {
    function FormFormatService(http) {
        this.http = http;
    }
    /**
     * @param {?} url
     * @param {?=} params
     * @return {?}
     */
    FormFormatService.prototype.getData = /**
     * @param {?} url
     * @param {?=} params
     * @return {?}
     */
    function (url, params) {
        /** @type {?} */
        var customData = params.customData;
        /** @type {?} */
        var index = params.pageIndex || 1;
        /** @type {?} */
        var size = params.pageSize || 20;
        /** @type {?} */
        var searchValue = JSON.parse(params.searchValue);
        /** @type {?} */
        var terminal = customData.tab === 'web' ? 'PC' : 'Mobile';
        /** @type {?} */
        var query = '?param=';
        /** @type {?} */
        var filter = "{\"terminal\":\"" + terminal + "\"";
        if (searchValue.searchValue) {
            if (searchValue.searchField === '*') {
                filter += ",\"allColumns\":\"" + searchValue.searchValue + "\"";
            }
            else {
                filter += ",\"" + searchValue.searchField + "\":\"" + searchValue.searchValue + "\"";
            }
        }
        filter += '}';
        filter = encodeURIComponent(filter);
        query += (filter + ("&pageNum=" + index + "&pageSize=" + size));
        url += query;
        return this.http.get(url).pipe(map((/**
         * @param {?} re
         * @return {?}
         */
        function (re) { return ({
            items: re.content,
            total: re.totalElements,
            pageInfo: {
                pageIndex: re.pageable.pageNumber + 1,
                pageSize: re.pageable.pageSize,
                pageList: [20, 40, 60, 80, 100],
                enablePager: true
            }
        }); })));
    };
    FormFormatService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FormFormatService.ctorParameters = function () { return [
        { type: HttpService }
    ]; };
    return FormFormatService;
}());
export { FormFormatService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FormFormatService.prototype.http;
}
