// @gsp-wf/wf-process-model/esm2015/lib/model/extension/assignment-strategy/AssignPerformerConfig.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ExtensionElement, BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
export class AssignPerformerConfig extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = AssignPerformerConfig.Clr_Type_ID;
        this.MultiSelect = true;
        this.Necessary = false;
        this.AllowGlobalAssign = false;
        this.Operator = 'currentUser';
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new AssignPerformerConfig(this.Model);
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "Operator", this.Operator);
        BpmnModelHelper.SetAttributeValue(obj, "MultiSelect", this.MultiSelect);
        BpmnModelHelper.SetAttributeValue(obj, "Necessary", this.Necessary);
        BpmnModelHelper.SetAttributeValue(obj, "AllowGlobalAssign", this.AllowGlobalAssign);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.Operator = obj.Operator;
        this.MultiSelect = obj.MultiSelect;
        this.Necessary = obj.Necessary;
        this.AllowGlobalAssign = obj.AllowGlobalAssign;
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const assignPerformerConfig = new AssignPerformerConfig(this.Model);
        assignPerformerConfig.Operator = this.Operator;
        assignPerformerConfig.MultiSelect = this.MultiSelect;
        assignPerformerConfig.Necessary = this.Necessary;
        assignPerformerConfig.AllowGlobalAssign = this.AllowGlobalAssign;
        return assignPerformerConfig;
    }
}
AssignPerformerConfig.Clr_Type_ID = 'wf.AssignPerformerConfig';
if (false) {
    /** @type {?} */
    AssignPerformerConfig.Clr_Type_ID;
    /** @type {?} */
    AssignPerformerConfig.prototype.ClrTypeID;
    /** @type {?} */
    AssignPerformerConfig.prototype.Operator;
    /** @type {?} */
    AssignPerformerConfig.prototype.MultiSelect;
    /** @type {?} */
    AssignPerformerConfig.prototype.Necessary;
    /** @type {?} */
    AssignPerformerConfig.prototype.AllowGlobalAssign;
}
BpmnModelHelper.RegisterElementType(new AssignPerformerConfig(new BpmnModel()));
