// @gsp-wf/wf-process-model/esm2015/lib/model/extension/assignment-strategy/AssignmentStrategyDef.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement } from "@edp-pmf/bpmn-model";
export class AssignmentStrategyDef extends ExtensionElement {
    //undo操作临时clone方法
    /**
     * @param {?} model
     * @param {?=} id
     * @param {?=} description
     * @param {?=} expandGroups
     */
    constructor(model, id, description, expandGroups) {
        super(model);
        this.ClrTypeID = "wf.AssignmentStrategy";
        this.Id = id;
        this.Description = description;
        this.ExpandGroups = expandGroups;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.Id = null;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new AssignmentStrategyDef(this.Model);
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "Id", this.Id);
        BpmnModelHelper.SetAttributeValue(obj, "Description", this.Description);
        BpmnModelHelper.SetAttributeValue(obj, "ExpandGroups", this.ExpandGroups);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.Id = obj.Id;
        this.Description = obj.Description;
        this.ExpandGroups = obj.ExpandGroups;
    }
}
if (false) {
    /** @type {?} */
    AssignmentStrategyDef.prototype.ClrTypeID;
    /** @type {?} */
    AssignmentStrategyDef.prototype.Id;
    /** @type {?} */
    AssignmentStrategyDef.prototype.Description;
    /** @type {?} */
    AssignmentStrategyDef.prototype.ExpandGroups;
}
BpmnModelHelper.RegisterElementType(new AssignmentStrategyDef(new BpmnModel()));
