// @gsp-svc/cloudprint/esm5/lib/entitys/outputoptions/outputOptions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FileType } from './fileType';
import { PrintType } from '../enums/printtype';
import { PrintSetting } from './printsetting';
/**
 * 导出设置
 */
var /**
 * 导出设置
 */
OutputOptions = /** @class */ (function () {
    function OutputOptions() {
        /**
         * 是否后台任务打印
         */
        this.printJob = false;
        /**
         * 打印机名称
         */
        this.printerName = '';
        /**
         * 打印设置
         */
        this.printSetting = new PrintSetting();
        this.FileType = FileType.Pdf;
        this.printType = PrintType.Form;
    }
    return OutputOptions;
}());
/**
 * 导出设置
 */
export { OutputOptions };
if (false) {
    /**
     * 输出类型 打印/导出
     * @type {?}
     */
    OutputOptions.prototype.OutputType;
    /**
     * 导出文件类型(默认是Pdf)
     * @type {?}
     */
    OutputOptions.prototype.FileType;
    /**
     * 导出路径
     * @type {?}
     */
    OutputOptions.prototype.Path;
    /**
     * 设备Id
     * @type {?}
     */
    OutputOptions.prototype.DeviceId;
    /**
     * 打印类型
     * @type {?}
     */
    OutputOptions.prototype.printType;
    /**
     * 是否后台任务打印
     * @type {?}
     */
    OutputOptions.prototype.printJob;
    /**
     * 打印机名称
     * @type {?}
     */
    OutputOptions.prototype.printerName;
    /**
     * 打印设置
     * @type {?}
     */
    OutputOptions.prototype.printSetting;
}
