// @gsp-svc/file-viewer/esm5/lib/file-list/file-list/file-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Input, Injector, ChangeDetectorRef } from '@angular/core';
import { LoadingService } from '@farris/ui-loading';
import { FileListService } from '../../service/file-list.service';
import { ImageViewerComponent } from '../../image-viewer/image-viewer/image-viewer.component';
import { MessagerService } from '@farris/ui-messager';
import { DocViewerComponent } from '../../doc-viewer/doc-viewer/doc-viewer.component';
import { OfdViewerComponent } from '../../ofd-viewer/ofd-viewer/ofd-viewer.component';
import { GspFileViewType } from '../../file-list-mobile/entity/gspfileviewtype';
import { PdfViewerComponent } from '../../pdf-viewer/pdf-viewer/pdf-viewer.component';
import { PdfViewOptions } from '../../pdf-viewer/entitires/pdfviewoptions';
import { PptViewerComponent } from '../../ppt-viewer/ppt-viewer/ppt-viewer.component';
var FileListComponent = /** @class */ (function () {
    function FileListComponent(listService, msgService, loadService, injector) {
        this.listService = listService;
        this.msgService = msgService;
        this.loadService = loadService;
        this.injector = injector;
        this.formId = '123456';
        this.rootId = 'default-root';
        this.metadataIdList = [];
        this.index = 0;
        this.metadataId = "";
        this.fileList = [];
        this.isImage = true;
        this.isOfd = true;
        this.isPdf = true;
        this.isPpt = true;
        //srcList: string[] = [];
        this.isDocShow = false;
        this.isOfdShow = false;
        this.isPdfShow = false;
        this.isPptShow = false;
        this.isFileListOpend = false;
    }
    /**
     * @return {?}
     */
    FileListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var loading = this.loadService.show({
            message: '拼命的请求中'
        });
        if (this.metadataIdList.length > 0) {
            this.listService.getFileList(this.metadataIdList, this.rootId).subscribe((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                loading.close();
                if (res.error != null) {
                    _this.msgService.warning('获取表单附件信息失败,请稍后重试。');
                }
                else {
                    _this.fileList = res;
                    _this.getFileIcon();
                    _this.getIndex();
                    _this.fileChange();
                    if (_this.injector) {
                        /** @type {?} */
                        var changeDetector = _this.injector.get(ChangeDetectorRef);
                        changeDetector.detectChanges();
                    }
                }
            }));
        }
        else {
            this.listService.getFormFileList(this.formId, this.rootId).subscribe((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                loading.close();
                if (res.error != null) {
                    _this.msgService.warning('获取表单附件信息失败,请稍后重试。');
                }
                else {
                    _this.fileList = res;
                    _this.getFileIcon();
                    _this.getIndex();
                    _this.fileChange();
                    if (_this.injector) {
                        /** @type {?} */
                        var changeDetector = _this.injector.get(ChangeDetectorRef);
                        changeDetector.detectChanges();
                    }
                }
            }));
        }
    };
    /**
     * @return {?}
     */
    FileListComponent.prototype.getIndex = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.metadataId == null || this.metadataId == "") {
            this.index = 0;
        }
        else if (this.fileList.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.metadataId == _this.metadataId; })).length == 0) {
            this.index = 0;
        }
        else {
            this.index = this.fileList.indexOf(this.fileList.filter((/**
             * @param {?} x
             * @return {?}
             */
            function (x) { return x.metadataId == _this.metadataId; }))[0]);
        }
    };
    /**
     * @return {?}
     */
    FileListComponent.prototype.fileChange = /**
     * @return {?}
     */
    function () {
        this.file = this.fileList[this.index];
        //this.file=src;
        this.isImage = this.file.fileType == GspFileViewType.Image || this.file.fileType == 'Image';
        this.isOfd = this.file.fileType == GspFileViewType.Ofd || this.file.fileType == 'Ofd';
        this.isPdf = this.file.fileType == GspFileViewType.Pdf || this.file.fileType == 'Pdf';
        this.isPpt = this.file.fileType == GspFileViewType.Ppt || this.file.fileType == 'Ppt';
        if (this.isImage) {
            this.isDocShow = false;
            this.isOfdShow = false; //this.imageChange(this.file.fileUrl);
            this.isPptShow = false;
            return;
        }
        else if (this.isOfd) {
            //this.file.fileUrl="/platform/runtime/dfs/images/file.jpg";
            if (this.isOfdShow)
                this.ofdChange(this.file.filePath);
            else {
                this.isDocShow = false;
                this.isOfdShow = true;
                this.isPdfShow = false;
                this.isPptShow = false;
            }
        }
        else if (this.isPdf) {
            if (this.isPdfShow)
                this.pdfChange(this.file.filePath);
            else {
                this.isDocShow = false;
                this.isOfdShow = false;
                this.isPdfShow = true;
                this.isPptShow = false;
            }
        }
        else if (this.isPpt) {
            if (this.isPptShow)
                this.pptChange(this.file.filePath);
            else {
                this.isDocShow = false;
                this.isOfdShow = false;
                this.isPdfShow = false;
                this.isPptShow = true;
            }
        }
        else {
            if (this.isDocShow)
                this.docChange(this.file.filePath);
            else {
                this.isDocShow = true;
                this.isOfdShow = false;
                this.isPdfShow = false;
                this.isPptShow = false;
            }
        }
    };
    /**
     * @param {?} src
     * @return {?}
     */
    FileListComponent.prototype.listBackColor = /**
     * @param {?} src
     * @return {?}
     */
    function (src) {
        if (this.file == src)
            return { "background-color": "#E6F7FF" };
    };
    /**
     * @param {?} src
     * @return {?}
     */
    FileListComponent.prototype.fileSelect = /**
     * @param {?} src
     * @return {?}
     */
    function (src) {
        // this.file==src;
        this.index = this.fileList.indexOf(src);
        this.fileChange();
        //this.imageViewer.changeFile(this.index);
    };
    /**
     * @param {?} url
     * @return {?}
     */
    FileListComponent.prototype.imageChange = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.imageUrl = url;
        this.imageViewer.imageChange(url);
    };
    /**
     * @param {?} filePath
     * @return {?}
     */
    FileListComponent.prototype.docChange = /**
     * @param {?} filePath
     * @return {?}
     */
    function (filePath) {
        this.docPath = filePath;
        this.docViewer.docChange(filePath);
    };
    /**
     * @param {?} filePath
     * @return {?}
     */
    FileListComponent.prototype.pdfChange = /**
     * @param {?} filePath
     * @return {?}
     */
    function (filePath) {
        //this.docPath = filePath;
        this.pdfViewer.pdfChange(filePath);
    };
    /**
     * @param {?} filePath
     * @return {?}
     */
    FileListComponent.prototype.pptChange = /**
     * @param {?} filePath
     * @return {?}
     */
    function (filePath) {
        //this.docPath = filePath;
        this.pptViewer.pptChange(filePath);
    };
    /**
     * @param {?} filePath
     * @return {?}
     */
    FileListComponent.prototype.ofdChange = /**
     * @param {?} filePath
     * @return {?}
     */
    function (filePath) {
        this.ofdViewer.fileChange(filePath);
    };
    /**
     * @return {?}
     */
    FileListComponent.prototype.getFileIcon = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var elementIsImage = true;
        /** @type {?} */
        var elementIsOfd = false;
        this.fileList.forEach((/**
         * @param {?} element
         * @return {?}
         */
        function (element) {
            elementIsImage = element.fileType == GspFileViewType.Image || element.fileType == 'Image';
            elementIsOfd = element.fileType == GspFileViewType.Ofd || element.fileType == 'Ofd';
            if (!elementIsImage && !elementIsOfd) {
                if (element.fileName.substring(element.fileName.lastIndexOf(".")).includes("doc") || element.fileName.substring(element.fileName.lastIndexOf(".")).includes("docx") || element.fileName.substring(element.fileName.lastIndexOf(".")).includes("wps"))
                    element.fileUrl = "/platform/runtime/dfs/images/word.png";
                else if (element.fileName.substring(element.fileName.lastIndexOf(".")).includes("xls") || element.fileName.substring(element.fileName.lastIndexOf(".")).includes("xlsx") || element.fileName.substring(element.fileName.lastIndexOf(".")).includes("et"))
                    element.fileUrl = "/platform/runtime/dfs/images/excel.png";
                else if (element.fileName.substring(element.fileName.lastIndexOf(".")).includes("ppt") || element.fileName.substring(element.fileName.lastIndexOf(".")).includes("pptx") || element.fileName.substring(element.fileName.lastIndexOf(".")).includes("dps"))
                    element.fileUrl = "/platform/runtime/dfs/images/ppt.png";
                else if (element.fileName.substring(element.fileName.lastIndexOf(".")).includes("pdf"))
                    element.fileUrl = "/platform/runtime/dfs/images/pdf.png";
                else if (element.fileName.substring(element.fileName.lastIndexOf(".")).includes("txt"))
                    element.fileUrl = "/platform/runtime/dfs/images/txt.png";
                else
                    element.fileUrl = "/platform/runtime/dfs/images/file.png";
            }
            if (elementIsOfd)
                element.fileUrl = "/platform/runtime/dfs/images/file.png";
        }));
    };
    /**
     * @return {?}
     */
    FileListComponent.prototype.isShowFileList = /**
     * @return {?}
     */
    function () {
        this.isFileListOpend = !this.isFileListOpend;
    };
    /**
     * @return {?}
     */
    FileListComponent.prototype.download = /**
     * @return {?}
     */
    function () {
        window.open(this.file.downloadUrl);
    };
    /**
     * @return {?}
     */
    FileListComponent.prototype.zoomIn = /**
     * @return {?}
     */
    function () {
        if (this.isImage) {
            this.imageViewer.zoomIn();
        }
        else if (this.isOfd) {
        }
        else {
            this.docViewer.zoomIn();
        }
    };
    /**
     * @return {?}
     */
    FileListComponent.prototype.zoomOut = /**
     * @return {?}
     */
    function () {
        if (this.isImage) {
            this.imageViewer.zoomOut();
        }
        else if (this.isOfd) {
        }
        else {
            this.docViewer.zoomOut();
        }
    };
    /**
     * @return {?}
     */
    FileListComponent.prototype.zoomReset = /**
     * @return {?}
     */
    function () {
        if (this.isImage) {
            this.imageViewer.zoomReset();
        }
        else if (this.isOfd) {
        }
        else {
            this.docViewer.zoomReset();
        }
    };
    /**
     * @return {?}
     */
    FileListComponent.prototype.rotateClockwise = /**
     * @return {?}
     */
    function () {
        if (this.isImage) {
            this.imageViewer.rotateClockwise();
        }
        else if (this.isOfd) {
        }
        else {
            this.docViewer.rotateClockwise();
        }
    };
    /**
     * @return {?}
     */
    FileListComponent.prototype.rotateCounterClockwise = /**
     * @return {?}
     */
    function () {
        if (this.isImage) {
            this.imageViewer.rotateCounterClockwise();
        }
        else if (this.isOfd) {
        }
        else {
            this.docViewer.rotateCounterClockwise();
        }
    };
    FileListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-file-list',
                    template: "<div class=\"f-utils-absolute-all\">\n    <kendo-splitter style=\"height:100%;width:100%\" orientation=\"vertical\">\n        <kendo-splitter-pane [collapsible]=\"true\" size=\"50px\" [resizable]='false'>\n\n            <div style=\"width: 100%\">\n                <div class=\"btn btn-secondary upload_button\" (click)=\"isShowFileList()\" style=\"left:40px;border:0\">\n                    <img [src]=\"'/platform/runtime/dfs/images/discount.png'\" class=\"upload_button\" style=\"top:0;left:0\" title=\"'\u6587\u6863\u5217\u8868'\">\n                </div>\n                <div class=\"btn btn-secondary upload_button\" (click)=\"download()\" style=\"left:90px;border:0\">\n                    <img [src]=\"'/platform/runtime/dfs/images/download.png'\" class=\"upload_button\" style=\"top:0;left:0\" title=\"'\u4E0B\u8F7D'\">\n                </div>\n                <div *ngIf=\"!isOfd && !isPdf && !isPpt\" class=\"btn btn-secondary upload_button\" (click)=\"zoomIn()\" style=\"left:50%; margin-left: -12.5px;;border:0\">\n                    <img [src]=\"'/platform/runtime/dfs/images/zoomin.png'\" class=\"upload_button\" style=\"top:0;left:0\" title=\"'\u653E\u5927'\">\n                </div>\n                <div *ngIf=\"!isOfd && !isPdf && !isPpt\" class=\"btn btn-secondary upload_button\" (click)=\"zoomOut()\" style=\"left:50%; margin-left: -62.5px;;border:0\">\n                    <img [src]=\"'/platform/runtime/dfs/images/zoomout.png'\" class=\"upload_button\" style=\"top:0;left:0\" title=\"'\u7F29\u5C0F'\">\n                </div>\n                <div *ngIf=\"!isOfd && !isPdf && !isPpt\" class=\"btn btn-secondary upload_button\" (click)=\"zoomReset()\" style=\"left:50%; margin-left: 37.5px;;border:0\">\n                    <img [src]=\"'/platform/runtime/dfs/images/zoomreset.png'\" class=\"upload_button\" style=\"top:2.5px;left:2.5px;width:20px;height:20px\" title=\"'\u91CD\u7F6E'\">\n                </div>\n                <div *ngIf=\"!isOfd && !isPdf && !isPpt\" class=\"btn btn-secondary upload_button\" (click)=\"rotateClockwise()\" style=\"right:40px;border:0\">\n                    <img [src]=\"'/platform/runtime/dfs/images/rotate.png'\" class=\"upload_button\" style=\"top:0;left:0\" title=\"'\u987A\u65F6\u9488\u65CB\u8F6C'\">\n                </div>\n                <div *ngIf=\"!isOfd && !isPdf && !isPpt\" class=\"btn btn-secondary upload_button\" (click)=\"rotateCounterClockwise()\" style=\"right:90px;border:0\">\n                    <img [src]=\"'/platform/runtime/dfs/images/conrotate.png'\" class=\"upload_button\" style=\"top:0;left:0\" title=\"'\u9006\u65F6\u9488\u65CB\u8F6C'\">\n                </div>\n            </div>\n\n        </kendo-splitter-pane>\n        <kendo-splitter-pane>\n            <kendo-splitter style=\"height:100%;width:100%\">\n                <kendo-splitter-pane [collapsible]=\"true\" size=\"200px\" [collapsed]=\"isFileListOpend\" [resizable]='false' [scrollable]='true' style=\"overflow-x:hidden\">\n                    <ul *ngFor=\"let src of fileList\" class=\"pic_list\" [ngStyle]=\"listBackColor(src)\">\n                        <a (click)=\"fileSelect(src)\">\n                            <div>\n                                <img [src]='src.fileUrl'>\n                            </div>\n                            <p>{{src.fileName}}</p>\n                        </a>\n                    </ul>\n                </kendo-splitter-pane>\n                <kendo-splitter-pane id=\"paneContent\" [scrollable]='true'>\n                    <div *ngIf='isImage&&!isOfd&&!isPdf&&!isPpt'>\n                        <app-image-viewer [imageUrl]=\"file.fileUrl\"></app-image-viewer>\n                    </div>\n                    <div *ngIf='!isImage&&!isOfd&&!isPdf&&!isPpt'>\n                        <app-doc-viewer [filePath]=\"file.filePath\"></app-doc-viewer>\n                    </div>\n                    <div *ngIf='!isImage&&isOfd&&!isPdf&&!isPpt'>\n                        <app-ofd-viewer [filePath]=\"file.filePath\"></app-ofd-viewer>\n                    </div>\n                    <div *ngIf='!isImage&&!isOfd&&!isPdf&&isPpt'>\n                        <app-ppt-viewer [pptUrl]=\"file.filePath\"></app-ppt-viewer>\n                    </div>\n                    <div *ngIf='!isImage&&!isOfd&&isPdf&&!isPpt' class=\"f-utils-fill\" style=\"height: 100%\">\n                        <app-pdf-viewer [pdfUrl]=\"file.filePath\" [viewOptions]=\"viewOptions\"></app-pdf-viewer>\n                    </div>\n                </kendo-splitter-pane>\n            </kendo-splitter>\n        </kendo-splitter-pane>\n    </kendo-splitter>\n</div>",
                    styles: [".button-group{width:100%;height:50px;display:table-cell}.button-group button{width:25px;height:25px;vertical-align:middle;position:absolute;top:12.5px}.button-group button img{width:25px;height:25px;vertical-align:middle}.pic_list{border:1px solid #ddd;padding:0;list-style:none;width:200px;height:200px;text-align:center}.pic_list a div{display:table-cell;vertical-align:middle;text-align:center;width:120px;height:140px;margin:20px 40px 10px}.pic_list a div img{vertical-align:middle;text-align:center;max-width:120px;max-height:140px}.pic_list a p{width:170px;height:40px;font-size:14px;text-align:center;overflow:hidden;text-overflow:ellipsis;word-wrap:break-word;word-break:break-all;white-space:nowrap;margin:10px 15px 0}.upload_button{width:25px;height:25px;vertical-align:middle;position:absolute;top:12.5px}"]
                }] }
    ];
    /** @nocollapse */
    FileListComponent.ctorParameters = function () { return [
        { type: FileListService },
        { type: MessagerService },
        { type: LoadingService },
        { type: Injector }
    ]; };
    FileListComponent.propDecorators = {
        formId: [{ type: Input }],
        rootId: [{ type: Input }],
        metadataIdList: [{ type: Input }],
        index: [{ type: Input }],
        metadataId: [{ type: Input }],
        viewOptions: [{ type: Input }],
        imageViewer: [{ type: ViewChild, args: [ImageViewerComponent,] }],
        docViewer: [{ type: ViewChild, args: [DocViewerComponent,] }],
        ofdViewer: [{ type: ViewChild, args: [OfdViewerComponent,] }],
        pdfViewer: [{ type: ViewChild, args: [PdfViewerComponent,] }],
        pptViewer: [{ type: ViewChild, args: [PptViewerComponent,] }]
    };
    return FileListComponent;
}());
export { FileListComponent };
if (false) {
    /** @type {?} */
    FileListComponent.prototype.formId;
    /** @type {?} */
    FileListComponent.prototype.rootId;
    /** @type {?} */
    FileListComponent.prototype.metadataIdList;
    /** @type {?} */
    FileListComponent.prototype.index;
    /** @type {?} */
    FileListComponent.prototype.metadataId;
    /** @type {?} */
    FileListComponent.prototype.viewOptions;
    /** @type {?} */
    FileListComponent.prototype.fileList;
    /** @type {?} */
    FileListComponent.prototype.file;
    /** @type {?} */
    FileListComponent.prototype.fileConent;
    /** @type {?} */
    FileListComponent.prototype.isImage;
    /** @type {?} */
    FileListComponent.prototype.isOfd;
    /** @type {?} */
    FileListComponent.prototype.isPdf;
    /** @type {?} */
    FileListComponent.prototype.isPpt;
    /** @type {?} */
    FileListComponent.prototype.imageUrl;
    /** @type {?} */
    FileListComponent.prototype.docPath;
    /** @type {?} */
    FileListComponent.prototype.isDocShow;
    /** @type {?} */
    FileListComponent.prototype.isOfdShow;
    /** @type {?} */
    FileListComponent.prototype.isPdfShow;
    /** @type {?} */
    FileListComponent.prototype.isPptShow;
    /** @type {?} */
    FileListComponent.prototype.isFileListOpend;
    /** @type {?} */
    FileListComponent.prototype.imageViewer;
    /** @type {?} */
    FileListComponent.prototype.docViewer;
    /** @type {?} */
    FileListComponent.prototype.ofdViewer;
    /** @type {?} */
    FileListComponent.prototype.pdfViewer;
    /** @type {?} */
    FileListComponent.prototype.pptViewer;
    /**
     * @type {?}
     * @private
     */
    FileListComponent.prototype.listService;
    /**
     * @type {?}
     * @private
     */
    FileListComponent.prototype.msgService;
    /**
     * @type {?}
     * @private
     */
    FileListComponent.prototype.loadService;
    /**
     * @type {?}
     * @private
     */
    FileListComponent.prototype.injector;
}
