// @gspwidget/widget-devkit/esm2015/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export { Widget } from './lib/widget-registry/widget.decorator';
export { WidgetRegistry } from './lib/widget-registry/widget-registry';
export { WidgetEvent, WidgetEventMap } from './lib/widget-registry/widget-event.decorator';
export { WidgetEventService } from './lib/widget-registry/widget-event.service';
export { WidgetPropertyService } from './lib/widget-registry/widget-property.service';
export { Property, WidgetPropertyMap } from './lib/widget-registry/widget-property.decorator';
/** widget-title-bar component */
export { WidgetTitleBarModule } from './lib/widget-title-bar/widget-title-bar.module';
export { WidgetTitleBarComponent } from './lib/widget-title-bar/widget-title-bar.component';
export { WidgetTitleBarSelectComponent } from './lib/widget-title-bar/widget-title-bar-select/widget-title-bar-select.component';
export { WidgetBase } from './lib/WidgetBase';
export { resolveTemplate, JumpTypes, jumpTypeOption, funcOption, dataJumpOption, datalessJumpOption, CommonPropertyOptions, filterYData, echartsGradient, tempLabel, EchartsOptions, chartsConsts, getChartColorsPure, getChartColorsGradient, getChartColorsGradientOne, EchartsPropertyOptions, tempMap } from './lib/preset/preset.index';
export { WidgetDataService } from './lib/data.service';
export { WidgetInstanceInfo } from './lib/widget-instance-info';
export { WidgetDevkitModule } from './lib/widget-devkit.module';
export { DASHBOARD_API, WidgetDevkitService } from './lib/widget-devkit.service';
export { PropertyTypes, HelpStructure } from './types';
export { CHART_TABLE_FUNC_ID } from './lib/util';
export { getObj, ifFsDateRange, isFsFlat, FSLastTimeItem, FSRangeDefault, FSDate, FSEnum, FSEnumItem, FilterSchema, FILTER_PROP_NAME, frontendFilterProp } from './lib/filter';
