// @farris/extend-fileupload-adapt-unifile/esm2015/lib/ffileupload-adapt-unifile.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ffileupload-adapt-unifile.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, Optional } from "@angular/core";
import { UploadServerService, UploadStatus, } from "@farris/extend-file-upload";
import { Observable, forkJoin, of, EMPTY, concat, Subject } from "rxjs";
import { UploadService, GspFormUploadListEntity, GspFormRemoveListEntity, GspFormDocInfo, GspFormUploadEntity, OperatingModes, } from "@gsp-svc/formdoc-upload";
import { FFileUploadAdaptUnifileConfigService, } from "./ffileupload-adapt-unifile.config";
import { catchError, delay, expand, last, map, switchMap, tap, } from "rxjs/operators";
import { LocaleService } from "@farris/ui-locale";
import * as i0 from "@angular/core";
import * as i1 from "@gsp-svc/formdoc-upload";
import * as i2 from "./ffileupload-adapt-unifile.config";
/**
 * @record
 */
export function UploadServerConfig() { }
if (false) {
    /** @type {?} */
    UploadServerConfig.prototype.sameNameAllowed;
    /** @type {?} */
    UploadServerConfig.prototype.maxFileSize;
    /** @type {?} */
    UploadServerConfig.prototype.allowedMultiUpload;
    /** @type {?|undefined} */
    UploadServerConfig.prototype.uploadedFileInfoList;
    /** @type {?|undefined} */
    UploadServerConfig.prototype.oldUploadedFileList;
}
export class FfileuploadAdaptUnifileService extends UploadServerService {
    /**
     * @param {?} uploadSer
     * @param {?} configSer
     * @param {?} inject
     */
    constructor(uploadSer, configSer, inject) {
        super();
        this.uploadSer = uploadSer;
        this.configSer = configSer;
        this.inject = inject;
        // 暂时用于简单合并
        this.bufferSize = 1024 * 1024;
        this.uploadedChunk = {};
        this.fileTotalChunk = {};
        this.serverConfig = null;
        this.extendData = this.configSer.getConfig();
        // 开始创建多语言服务
        /** @type {?} */
        let hasLocalSer = this.configSer.setLocalState(null);
        if (this.inject && !hasLocalSer) {
            // 如果存在inject，再次创建，inject是在某个版不能之后才出现的
            this.configSer.setLocalState(this.inject.get(LocaleService, null));
        }
    }
    /**
     * 因为被附件组件作为依赖类，导致此处获取不到最新的基路径值
     * @private
     * @return {?}
     */
    getPrefixStr() {
        return this.configSer.getBasePath();
    }
    /**
     * @private
     * @return {?}
     */
    uuid() {
        /** @type {?} */
        let S4 = (/**
         * @return {?}
         */
        function () {
            return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
        });
        /** @type {?} */
        let nGuid = S4() + S4() + '-' + S4() + '-' + S4() + '-' + S4() + '-' + S4() + S4() + S4();
        return nGuid;
    }
    /**
     * 删除文件
     * @param {?} files
     * @param {?} event
     * @param {?} extendService
     * @return {?}
     */
    remove(files, event, extendService) {
        return this.innerRemoveList(files, event, extendService);
    }
    /**
     * @private
     * @param {?} files
     * @param {?} event
     * @param {?=} extendService
     * @return {?}
     */
    innerRemoveList(files, event, extendService = null) {
        return new Observable((/**
         * @param {?} observer
         * @return {?}
         */
        (observer) => {
            /** @type {?} */
            let removeFileInfo = new GspFormRemoveListEntity();
            /** @type {?} */
            let metadataIdList = [];
            files.forEach((/**
             * @param {?} fileItem
             * @return {?}
             */
            (fileItem) => {
                if (fileItem["response"]) {
                    metadataIdList.push(fileItem["response"]["metadataId"]);
                }
            }));
            removeFileInfo.mode = this.getFinallyConfig("mode", extendService);
            /** @type {?} */
            let rootId = this.getFinallyConfig("rootId", extendService);
            removeFileInfo.metadataIdList = [].concat(metadataIdList);
            this.uploadSer.removeList(metadataIdList, rootId).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            (result) => {
                // 没有需要整合的服务器端返回数据
                observer.next({ type: "removed", files: files });
            }), (/**
             * @param {?} error
             * @return {?}
             */
            (error) => {
                observer.error(error);
                observer.complete();
            }), (/**
             * @return {?}
             */
            () => {
                observer.complete();
            }));
        }));
    }
    /**
     * 上传文件
     * @param {?} files
     * @param {?} event
     * @param {?} extendService
     * @return {?}
     */
    upload(files, event, extendService) {
        this.formId = this.getFinallyConfig("formId", extendService);
        this.rootId = this.getFinallyConfig("rootId", extendService);
        if (event.type == "sliceUpload") {
            /** @type {?} */
            const up$ = (/**
             * @return {?}
             */
            () => {
                if (this.serverConfig && this.serverConfig.allowedMultiUpload) {
                    if (!this.serverConfig.sameNameAllowed) {
                        return this.uploadSer.getUploadedFileInfoList(this.formId, this.rootId).pipe(switchMap((/**
                         * @param {?} res
                         * @return {?}
                         */
                        (res) => {
                            if (res != null && res.error == null) {
                                this.serverConfig.uploadedFileInfoList = res;
                                this.serverConfig.oldUploadedFileList = JSON.parse(JSON.stringify(res));
                                return this.uploadBigFile2(files, event, extendService);
                            }
                            return of(res);
                        })));
                    }
                    else {
                        return this.uploadBigFile2(files, event, extendService);
                    }
                }
                else {
                    return this.uploadBigFile(files, event, extendService);
                }
            });
            if (!this.serverConfig) {
                return this.getServerConfig(this.rootId).pipe(switchMap((/**
                 * @param {?} res
                 * @return {?}
                 */
                (res) => {
                    if (res != null && res.error == null) {
                        this.serverConfig = {
                            sameNameAllowed: res.sameNameAllowed,
                            allowedMultiUpload: res.allowedMultiUpload,
                            maxFileSize: parseInt(JSON.parse(res.validateConfiguration).maxFileSize)
                        };
                    }
                    return up$();
                })));
            }
            return up$();
        }
        return this.innerUploadList(files, event, extendService);
    }
    /**
     * 内部上传方式
     * @private
     * @param {?} files
     * @param {?} event
     * @param {?} extendService
     * @return {?}
     */
    innerUploadList(files, event, extendService) {
        return new Observable((/**
         * @param {?} observer
         * @return {?}
         */
        (observer) => {
            /** @type {?} */
            let uploadInfo = new GspFormUploadListEntity();
            uploadInfo.formId = this.getFinallyConfig("formId", extendService);
            uploadInfo.mode = this.getFinallyConfig("mode", extendService);
            /** @type {?} */
            let rootId = this.getFinallyConfig("rootId", extendService);
            uploadInfo.docInfoList = [];
            /** @type {?} */
            let readerObserable = [];
            files.forEach((/**
             * @param {?} fileItem
             * @return {?}
             */
            (fileItem) => {
                /** @type {?} */
                let sub = new Observable((/**
                 * @param {?} obser
                 * @return {?}
                 */
                (obser) => {
                    /** @type {?} */
                    let reader = new FileReader();
                    // 暂不处理异常
                    reader.readAsBinaryString(fileItem.nativeFile);
                    reader.onload = (/**
                     * @param {?} e
                     * @return {?}
                     */
                    (e) => {
                        /** @type {?} */
                        let info = (/** @type {?} */ ({ fileName: "", fileContent: "" }));
                        info["fileName"] = fileItem.name;
                        info["fileContent"] = btoa(reader.result.toString());
                        // 如果data里有数据，并且属性是配置的扩展属性
                        if (event.hasOwnProperty("data") &&
                            event["data"] &&
                            event["data"].hasOwnProperty("extProperty")) {
                            //
                            info["extProperty"] = event["data"]["extProperty"];
                        }
                        uploadInfo.docInfoList.push(info);
                        obser.next();
                        obser.complete();
                    });
                }));
                readerObserable.push(sub);
            }));
            forkJoin(readerObserable).subscribe((/**
             * @param {?} datas
             * @return {?}
             */
            (datas) => {
                //     metadataId: string;
                //  fileName: string;
                this.uploadSer.uploadList(uploadInfo, rootId).subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                (result) => {
                    if (result.error) {
                        // 此接口一旦出现问题，所有附件都上传不通过
                        observer.error(this.errorInfoFormat(result.error, files));
                        observer.complete();
                        return;
                    }
                    // 合并返回结果到此处
                    result.forEach((/**
                     * @param {?} item
                     * @return {?}
                     */
                    (item) => {
                        /** @type {?} */
                        var findIndex = this.findFileIndexByFileName(files, item.fileName);
                        if (findIndex > -1) {
                            files[findIndex]["response"] = item;
                            files[findIndex]["progress"]["status"] = UploadStatus.Done;
                        }
                    }));
                    observer.next({ type: "done", files: files });
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                (error) => {
                    observer.error(this.errorInfoFormat(error, files));
                    observer.complete();
                }), (/**
                 * @return {?}
                 */
                () => {
                    observer.complete();
                }));
            }));
        }));
    }
    /**
     * 分片加载
     * @param {?} file
     * @param {?} event
     * @param {?} extendService
     * @return {?}
     */
    multipartUpload(file, event, extendService) {
        return new Observable((/**
         * @param {?} observer
         * @return {?}
         */
        (observer) => {
            // let fileInfo: FileInfo;
            /** @type {?} */
            let uuid = this.uuid();
            // let uuid = UUID.UUID();
            //reader.readAsBinaryString(selectedFile);
            //reader.onload = (e) => {
            /** @type {?} */
            let fileName = file.name;
            // fileInfo = {
            //   id: uuid,
            //   name: fileName,
            //   type: file.type,
            //   source: "",
            //   size: this.formatFileSize(file.size),
            //   fileSize: file.size,
            //   picListDisplayName: this.getMultipartDisplayName(fileName),
            //   queueListDisplayName: fileName,
            //   //content: reader.result.toString(),
            //   hasUploaded: false,
            //   mouseOn: false,
            //   selectd: false,
            //   isUploading: true,
            //   uploadResult: false,
            //   uploadProcess: 0,
            //   errorMessage: "",
            //   extensionDropListId: 0,
            //   extensionName: ""
            // }
            /** @type {?} */
            let chunkTotal = Math.ceil(file.size / this.bufferSize);
            // 标记当前附件总的分片数
            this.fileTotalChunk[uuid] = chunkTotal;
            /** @type {?} */
            let chunkIndex = 0;
            // 标记当前附件成功的分片数
            this.uploadedChunk[uuid] = 0;
            while (chunkIndex < chunkTotal) {
                /** @type {?} */
                let uploadInfo = new GspFormUploadEntity();
                uploadInfo.mode = OperatingModes.Temp;
                uploadInfo.formId = this.getFinallyConfig("formId", extendService);
                uploadInfo.rootId = this.getFinallyConfig("rootId", extendService);
                /** @type {?} */
                let docInfo = new GspFormDocInfo();
                docInfo.fileName = fileName;
                docInfo.metadataId = uuid;
                docInfo.total = chunkTotal;
                // 如果data里有数据，并且属性是配置的扩展属性
                if (event.hasOwnProperty("data") &&
                    event["data"] &&
                    event["data"].hasOwnProperty("extProperty")) {
                    docInfo["extProperty"] = (/** @type {?} */ (event["data"]["extProperty"]));
                }
                // 下一个文件大小
                /** @type {?} */
                let nextSize = Math.min((chunkIndex + 1) * this.bufferSize, file.size);
                // File类型
                /** @type {?} */
                let fileData = file["nativeFile"].slice(chunkIndex * this.bufferSize, nextSize);
                // 读取文件
                /** @type {?} */
                let reader = new FileReader();
                reader.readAsBinaryString(fileData);
                /** @type {?} */
                let innerIndex = chunkIndex;
                reader.onload = (/**
                 * @return {?}
                 */
                () => {
                    // 附件上传
                    docInfo.fileContent = btoa(reader.result.toString());
                    docInfo.index = innerIndex;
                    uploadInfo.docInfo = docInfo;
                    /** @type {?} */
                    let info = uploadInfo;
                    this.uploadSer.uploadFile(info).subscribe((/**
                     * @param {?} result
                     * @return {?}
                     */
                    (result) => {
                        if (result && result.error) {
                            // 分片上传有返回结果就是报错了
                            observer.error(this.errorInfoFormat(result.error, [file]));
                            observer.complete();
                            return;
                        }
                        else {
                            this.uploadedChunk[uuid]++;
                            if (this.uploadedChunk[docInfo.metadataId] ==
                                this.fileTotalChunk[docInfo.metadataId]) {
                                // 判断是所有分片都上传完成
                                file.progress = {
                                    status: UploadStatus.Done,
                                    data: {
                                        percentage: 100,
                                    },
                                };
                                file.response = docInfo;
                                delete this.uploadedChunk[uuid];
                                delete this.fileTotalChunk[uuid];
                                observer.next({ type: "done", files: [file] });
                                observer.complete();
                            }
                            else {
                                // 分片上传正常
                                /** @type {?} */
                                const percentage = Number.parseInt(((this.uploadedChunk[uuid] / this.fileTotalChunk[uuid]) *
                                    100).toFixed(0));
                                // 判断未上传完成
                                file.progress = {
                                    status: UploadStatus.Uploading,
                                    data: {
                                        percentage: percentage,
                                    },
                                };
                                observer.next({ type: "uploading", files: [file] });
                            }
                        }
                    }), (/**
                     * @param {?} error
                     * @return {?}
                     */
                    (error) => {
                        delete this.uploadedChunk[uuid];
                        delete this.fileTotalChunk[uuid];
                        observer.error(this.errorInfoFormat(error, [file]));
                        observer.complete();
                    }));
                });
                chunkIndex = chunkIndex + 1;
            }
        }));
    }
    /**
     * 获取分片的附件大小
     * @private
     * @param {?} size
     * @return {?}
     */
    formatFileSize(size) {
        if (size < 102400)
            return (size / 1024).toFixed(1) + "K";
        else if (size < 1024 * 1024)
            return (size / 1024).toFixed(0) + "K";
        else if (size < 100 * 1024 * 1024)
            return (size / 1024 / 1024).toFixed(1) + "M";
        else if (size < 1024 * 1024 * 1024)
            return (size / 1024 / 1024).toFixed(0) + "M";
        else
            return (size / 1024 / 1024 / 1024).toFixed(1) + "G";
    }
    /**
     * 获取分片的 附件名称
     * @private
     * @param {?} fileName
     * @return {?}
     */
    getMultipartDisplayName(fileName) {
        if (fileName.length <= 10)
            return fileName;
        else {
            /** @type {?} */
            let headContent = fileName.substring(0, 2);
            /** @type {?} */
            let tailContent = fileName.substring(fileName.lastIndexOf(".") - 2);
            return headContent + "…" + tailContent;
        }
    }
    /**
     * @private
     * @param {?} error
     * @param {?} files
     * @return {?}
     */
    errorInfoFormat(error, files) {
        //
        /** @type {?} */
        let errorFiles = files.map((/**
         * @param {?} item
         * @return {?}
         */
        (item) => {
            return { id: item.id, name: item.name };
        }));
        /** @type {?} */
        let errorMsg = this.configSer.getLocalStr('uploadAdapt.uploadError');
        if (error) {
            return Object.assign(error, { files: errorFiles }, {
                message: error["Message"] || error["extensionMessage"] || errorMsg,
                type: "error",
            });
        }
        else {
            return Object.assign({ files: errorFiles }, { message: errorMsg, type: "error" });
        }
    }
    /**
     * 返回最终属性值
     * @private
     * @param {?} key
     * @param {?} extendValue
     * @return {?}
     */
    getFinallyConfig(key, extendValue) {
        if (extendValue && extendValue.hasOwnProperty(key)) {
            return extendValue[key];
        }
        return this.extendData[key];
    }
    /**
     * @private
     * @param {?} files
     * @param {?} fileName
     * @return {?}
     */
    findFileIndexByFileName(files, fileName) {
        // 按照肯定能找到文件来处理
        /** @type {?} */
        let findIndex = files.findIndex((/**
         * @param {?} file
         * @return {?}
         */
        (file) => file.name == fileName));
        return findIndex;
    }
    //#region 大文件分片上传
    /**
     * @param {?} files
     * @param {?} event
     * @param {?} extendService
     * @return {?}
     */
    uploadBigFile(files, event, extendService) {
        /** @type {?} */
        const uploadChunks$ = new Subject();
        /** @type {?} */
        const _files = files.map((/**
         * @param {?} n
         * @return {?}
         */
        (n) => {
            /** @type {?} */
            const chunks = this._getBigFileChunks(n, event, extendService);
            return Object.assign({}, n, chunks);
        }));
        /** @type {?} */
        const uploadchunks$ = _files.map((/**
         * @param {?} f
         * @return {?}
         */
        (f) => this.uploadChunks(f, uploadChunks$)));
        concat(...uploadchunks$).subscribe((/**
         * @param {?} f
         * @return {?}
         */
        (f) => {
            f.progress = {
                status: UploadStatus.Done,
                data: {
                    percentage: 100,
                },
            };
            f.response = JSON.parse(f.chunks[0].get("docInfo"));
            uploadChunks$.next({ type: "done", files: [f] });
        }), (/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            uploadChunks$.error(this.errorInfoFormat(error, files));
            uploadChunks$.complete();
        }), (/**
         * @return {?}
         */
        () => {
            uploadChunks$.complete();
        }));
        return uploadChunks$.asObservable();
    }
    /**
     * @private
     * @param {?} file
     * @return {?}
     */
    uploadFileChunk(file) {
        /** @type {?} */
        const url = `${this.getPrefixStr()}/api/runtime/dfs/v1.0/formdoc/slice`;
        return this.uploadSer["http"]["http"]
            .post(url, file.chunks[file.total])
            .pipe(switchMap((/**
         * @return {?}
         */
        () => {
            return of(file);
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            return of(error);
        })));
    }
    /**
     * @private
     * @param {?} file
     * @param {?} upChunk$
     * @return {?}
     */
    uploadChunks(file, upChunk$) {
        /** @type {?} */
        const doChunksRemain = (/**
         * @param {?} i
         * @return {?}
         */
        (i) => i > -1);
        /** @type {?} */
        const completeUpload = (/**
         * @param {?} f
         * @return {?}
         */
        (f) => of(f));
        /** @type {?} */
        const uploadChunk = (/**
         * @param {?} f
         * @return {?}
         */
        (f) => {
            return this.uploadFileChunk(f).pipe(delay(100), map((/**
             * @param {?} f
             * @return {?}
             */
            (f) => {
                // 块数为0 时，即为此文件的所有块均上传成功
                f.progress = {
                    status: UploadStatus.Uploading,
                    data: {
                        percentage: ((f.total / f.chunks.length) * 100).toFixed(0),
                    },
                };
                upChunk$.next({ type: "uploading", files: [f] });
                return f;
            })));
        });
        /** @type {?} */
        const createMultipartUpload = (/**
         * @param {?} f
         * @return {?}
         */
        (f) => of(f));
        return createMultipartUpload(file).pipe(
        // tap((f) => console.log("upload file", f.name)),
        expand((/**
         * @param {?} f
         * @return {?}
         */
        (f) => {
            if (doChunksRemain(--file.total)) {
                return uploadChunk(file);
            }
            return EMPTY;
        })), 
        // tap((f: any) => {
        //     console.log(`file: ${f.name} - remaining chunks: ${f.total}`);
        // }),
        last(), switchMap((/**
         * @param {?} res
         * @return {?}
         */
        (res) => {
            return completeUpload(file);
        })));
    }
    /**
     * @private
     * @param {?} file
     * @param {?} event
     * @param {?} extendService
     * @return {?}
     */
    _getBigFileChunks(file, event, extendService) {
        /** @type {?} */
        let uuid = this.uuid();
        /** @type {?} */
        const formId = this.getFinallyConfig("formId", extendService);
        /** @type {?} */
        const rootId = this.getFinallyConfig("rootId", extendService);
        /** @type {?} */
        let realChunkSize = (event.chunkSize || 1) * this.bufferSize;
        /** @type {?} */
        let chunkTotal = Math.ceil(file.size / realChunkSize);
        /** @type {?} */
        let chunkIndex = 0;
        /** @type {?} */
        const fileChunks = { chunks: [], total: chunkTotal };
        while (chunkIndex < chunkTotal) {
            /** @type {?} */
            let uploadInfo = new GspFormUploadEntity();
            uploadInfo.mode = OperatingModes.Temp;
            uploadInfo.formId = formId;
            uploadInfo.rootId = rootId;
            /** @type {?} */
            let docInfo = new GspFormDocInfo();
            docInfo.fileName = file.name;
            docInfo.metadataId = uuid;
            docInfo.total = chunkTotal;
            /** @type {?} */
            let nextSize = Math.min((chunkIndex + 1) * realChunkSize, file.size);
            // File类型
            /** @type {?} */
            let fileData = file["nativeFile"].slice(chunkIndex * realChunkSize, nextSize);
            docInfo.size = file.size;
            docInfo.index = chunkIndex;
            docInfo.fileContent = "";
            // 如果data里有数据，并且属性是配置的扩展属性
            if (event.hasOwnProperty("data") &&
                event["data"] &&
                event["data"].hasOwnProperty("extProperty")) {
                docInfo["extProperty"] = (/** @type {?} */ (event["data"]["extProperty"]));
            }
            uploadInfo.docInfo = docInfo;
            //封装请求formdata
            /** @type {?} */
            let formData = new FormData();
            formData.append("uploadInfo", JSON.stringify(uploadInfo));
            formData.append("docInfo", JSON.stringify(docInfo));
            formData.append("file", fileData);
            fileChunks.chunks.push(formData);
            chunkIndex = chunkIndex + 1;
        }
        return fileChunks;
    }
    /**
     * 获取跟目录的配置
     * @private
     * @param {?} rootId
     * @return {?}
     */
    getServerConfig(rootId) {
        return this.uploadSer.getUploadInfo(rootId);
    }
    //#endregion
    //#region 读取相关配置分片上传至云服务器
    /**
     * @private
     * @param {?} file
     * @param {?} uploadChunks$
     * @return {?}
     */
    unorderedUploadBigFile(file, uploadChunks$) {
        if (!file.partList) {
            file.partList = [];
        }
        this.initMultiUpload(file).subscribe((/**
         * @param {?} r
         * @return {?}
         */
        (r) => {
            if (r && r.error) {
                uploadChunks$.error(this.errorInfoFormat(r.error, [file]));
                uploadChunks$.complete();
                return;
            }
            file.chunks.forEach((/**
             * @param {?} n
             * @return {?}
             */
            n => {
                this.multiUpload(n, r).subscribe((/**
                 * @param {?} res
                 * @return {?}
                 */
                (res) => {
                    if (res && res.error) {
                        uploadChunks$.error(this.errorInfoFormat(res.error, [file]));
                        uploadChunks$.complete();
                        return;
                    }
                    file.progress = {
                        status: UploadStatus.Uploading,
                        data: {
                            percentage: (res.index / file.chunks.length * 100).toFixed(0)
                        }
                    };
                    uploadChunks$.next({ type: 'uploading', files: [file] });
                    if (res && res.result) {
                        file.partList.push(res.result);
                    }
                    if (file.partList.length === file.chunks.length) {
                        /** @type {?} */
                        let completeMultiUploadRequest = {};
                        completeMultiUploadRequest.metadataId = file.metadataId;
                        completeMultiUploadRequest.partList = file.partList;
                        completeMultiUploadRequest.rootId = this.rootId;
                        completeMultiUploadRequest.uploadId = file.uploadId;
                        completeMultiUploadRequest.size = file.size;
                        this.completeMultiUpload(completeMultiUploadRequest).subscribe((/**
                         * @param {?} cmur
                         * @return {?}
                         */
                        (cmur) => {
                            if (cmur && cmur.status != 200) {
                                uploadChunks$.error(this.errorInfoFormat(cmur.error, [file]));
                                uploadChunks$.complete();
                                return;
                            }
                            file.progress = {
                                status: UploadStatus.Done,
                                data: {
                                    percentage: 100
                                }
                            };
                            file.response = JSON.parse(file.chunks[0].get("docInfo"));
                            uploadChunks$.next({ type: 'done', files: [file] });
                        }), (/**
                         * @param {?} error
                         * @return {?}
                         */
                        error => {
                            uploadChunks$.error(this.errorInfoFormat(error, [file]));
                            uploadChunks$.complete();
                        }));
                    }
                }));
            }));
        }));
    }
    /**
     * @param {?} files
     * @param {?} event
     * @param {?} extendService
     * @return {?}
     */
    uploadBigFile2(files, event, extendService) {
        /** @type {?} */
        const uploadChunks$ = new Subject();
        /** @type {?} */
        const formId = this.getFinallyConfig('formId', extendService);
        /** @type {?} */
        const rootId = this.getFinallyConfig('rootId', extendService);
        /** @type {?} */
        const _files = files.map((/**
         * @param {?} n
         * @return {?}
         */
        n => {
            /** @type {?} */
            const chunks = this._getBigFileChunks2(n, event, extendService);
            return Object.assign({}, n, chunks, { formId, rootId });
        }));
        // 无序上传
        _files.forEach((/**
         * @param {?} n
         * @return {?}
         */
        n => {
            this.unorderedUploadBigFile(n, uploadChunks$);
        }));
        // 有序上传
        // const uploadchunks$: Observable<any>[] =  _files.map((f) => this.uploadChunks2(f, uploadChunks$));
        // concat(...uploadchunks$).pipe(
        //     switchMap((f: any) => {
        //         let completeMultiUploadRequest: any = {};
        //         completeMultiUploadRequest.metadataId = f.metadataId;
        //         completeMultiUploadRequest.partList = f.partList;
        //         completeMultiUploadRequest.rootId = this.rootId;
        //         completeMultiUploadRequest.uploadId = f.uploadId;
        //         completeMultiUploadRequest.size = f.size;
        //         return this.completeMultiUpload(completeMultiUploadRequest).pipe(
        //             map(() => f)
        //         );
        //     })
        // ).subscribe(
        //     (f: any) => {
        //         f.progress = {
        //             status: UploadStatus.Done,
        //             data:  {
        //                 percentage: 100
        //             }
        //         };
        //         f.response = JSON.parse(f.chunks[0].get("docInfo"));
        //         uploadChunks$.next({type: 'done', files:[f]});
        //     },
        //     error => {
        //         uploadChunks$.error(this.errorInfoFormat(error, files));
        //         uploadChunks$.complete();
        //     },
        //     () => {
        //         uploadChunks$.complete();
        //         // console.log("all files uploaded")
        //     }
        // );
        return uploadChunks$.asObservable();
    }
    /**
     * @private
     * @param {?} file
     * @return {?}
     */
    initMultiUpload(file) {
        /** @type {?} */
        let initRequest = {};
        initRequest.path = file.formId;
        initRequest.metadataId = this.uuid();
        initRequest.rootId = file.rootId;
        initRequest.size = file.size;
        initRequest.fileName = file.name;
        /** @type {?} */
        let url = `${this.getPrefixStr()}/api/runtime/dfs/v1.0/formdoc/multi/init`;
        return this.uploadSer['http']['http'].post(url, initRequest).pipe(tap((/**
         * @param {?} res
         * @return {?}
         */
        (res) => {
            file.metadataId = initRequest.metadataId;
            file.uploadId = res.uploadId;
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            return of(error);
        })), map((/**
         * @param {?} res
         * @return {?}
         */
        (res) => {
            res.metadataId = initRequest.metadataId;
            return res;
        })));
    }
    /**
     * @private
     * @param {?} request
     * @return {?}
     */
    completeMultiUpload(request) {
        /** @type {?} */
        let url = `${this.getPrefixStr()}/api/runtime/dfs/v1.0/formdoc/multi/complete`;
        return this.uploadSer['http']['http'].post(url, request).pipe(tap((/**
         * @param {?} res
         * @return {?}
         */
        (res) => { })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            return of(error);
        })));
    }
    /**
     * @private
     * @param {?} formdata
     * @param {?} r
     * @return {?}
     */
    multiUpload(formdata, r) {
        /** @type {?} */
        let url = `${this.getPrefixStr()}/api/runtime/dfs/v1.0/formdoc/multi/upload`;
        /** @type {?} */
        const mr = JSON.parse((/** @type {?} */ (formdata.get('multiRequest'))));
        mr.uploadId = r.uploadId;
        mr.metadataId = r.metadataId;
        formdata.set('multiRequest', JSON.stringify(mr));
        /** @type {?} */
        const di = JSON.parse((/** @type {?} */ (formdata.get('docInfo'))));
        di.metadataId = r.metadataId;
        formdata.set('docInfo', JSON.stringify(di));
        return this.uploadSer['http']['http'].post(url, formdata).pipe(tap((/**
         * @param {?} res
         * @return {?}
         */
        res => { })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            return of(error);
        })));
    }
    /**
     * @private
     * @param {?} file
     * @param {?} upChunk$
     * @return {?}
     */
    uploadChunks2(file, upChunk$) {
        /** @type {?} */
        const total = file.total;
        /** @type {?} */
        let t = -1;
        /** @type {?} */
        const doChunksRemain = (/**
         * @param {?} i
         * @return {?}
         */
        (i) => i < total);
        /** @type {?} */
        const completeUpload = (/**
         * @param {?} f
         * @return {?}
         */
        (f) => of(f));
        /** @type {?} */
        const uploadChunk = (/**
         * @param {?} f
         * @param {?} uploadId
         * @return {?}
         */
        (f, uploadId) => {
            return this.multiUpload(f, uploadId).pipe(delay(100), map((/**
             * @param {?} res
             * @return {?}
             */
            (res) => {
                // 块数为0 时，即为此文件的所有块均上传成功
                file.progress = {
                    status: UploadStatus.Uploading,
                    data: {
                        percentage: (res.index / file.chunks.length * 100).toFixed(0)
                    }
                };
                upChunk$.next({ type: 'uploading', files: [file] });
                if (res && res.result) {
                    if (!file.partList) {
                        file.partList = [];
                    }
                    file.partList.push(res.result);
                }
                return file;
            })));
        });
        /** @type {?} */
        const createMultipartUpload = (/**
         * @param {?} f
         * @return {?}
         */
        (f) => {
            return this.initMultiUpload(f);
            // return of(f)
        });
        return createMultipartUpload(file).pipe(
        // tap((f) => console.log("upload file", f.name)),
        expand((/**
         * @param {?} r
         * @return {?}
         */
        (r) => {
            if (doChunksRemain(++t)) {
                return uploadChunk(file.chunks[t], r);
            }
            return EMPTY;
        })), 
        // tap((f: any) => {
        //     console.log(`file: ${f.name} - remaining chunks: ${f.total}`);
        // }),
        last(), switchMap((/**
         * @param {?} res
         * @return {?}
         */
        (res) => {
            return completeUpload(file);
        })));
    }
    /**
     * @private
     * @param {?} file
     * @param {?} event
     * @param {?} extendService
     * @return {?}
     */
    _getBigFileChunks2(file, event, extendService) {
        /** @type {?} */
        let uuid = this.uuid();
        /** @type {?} */
        const _bifferSize = Math.max(event.chunkSize, 5) * this.bufferSize;
        /** @type {?} */
        let chunkTotal = Math.ceil(file.size / _bifferSize);
        /** @type {?} */
        let chunkIndex = 0;
        /** @type {?} */
        const fileChunks = { chunks: [], total: chunkTotal };
        /** @type {?} */
        const rootId = this.getFinallyConfig('rootId', extendService);
        while (chunkIndex < chunkTotal) {
            /** @type {?} */
            let nextSize = Math.min((chunkIndex + 1) * _bifferSize, file.size);
            // File类型
            /** @type {?} */
            let fileData = file['nativeFile'].slice(chunkIndex * _bifferSize, nextSize);
            /** @type {?} */
            let multiRequest = {};
            multiRequest.metadataId = uuid;
            multiRequest.rootId = rootId;
            // multiRequest.uploadId = uploadId;
            multiRequest.index = chunkIndex;
            multiRequest.total = chunkTotal;
            multiRequest.size = fileData.size;
            //封装请求formdata
            /** @type {?} */
            let formData = new FormData();
            formData.append('multiRequest', JSON.stringify(multiRequest));
            formData.append('file', fileData);
            /** @type {?} */
            let docInfo = new GspFormDocInfo();
            docInfo.fileName = file.name;
            docInfo.metadataId = uuid;
            docInfo.total = chunkTotal;
            docInfo.size = file.size;
            docInfo.index = chunkIndex;
            docInfo.fileContent = "";
            formData.append("docInfo", JSON.stringify(docInfo));
            // 如果data里有数据，并且属性是配置的扩展属性
            if (event.hasOwnProperty("data") &&
                event["data"] &&
                event["data"].hasOwnProperty("extProperty")) {
                docInfo["extProperty"] = (/** @type {?} */ (event["data"]["extProperty"]));
            }
            fileChunks.chunks.push(formData);
            chunkIndex = chunkIndex + 1;
        }
        return fileChunks;
    }
}
FfileuploadAdaptUnifileService.previous = 0;
FfileuploadAdaptUnifileService.decorators = [
    { type: Injectable, args: [{
                providedIn: "root",
            },] }
];
/** @nocollapse */
FfileuploadAdaptUnifileService.ctorParameters = () => [
    { type: UploadService },
    { type: FFileUploadAdaptUnifileConfigService },
    { type: Injector, decorators: [{ type: Optional }] }
];
/** @nocollapse */ FfileuploadAdaptUnifileService.ngInjectableDef = i0.defineInjectable({ factory: function FfileuploadAdaptUnifileService_Factory() { return new FfileuploadAdaptUnifileService(i0.inject(i1.UploadService), i0.inject(i2.FFileUploadAdaptUnifileConfigService), i0.inject(i0.INJECTOR, 8)); }, token: FfileuploadAdaptUnifileService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    FfileuploadAdaptUnifileService.previous;
    /** @type {?} */
    FfileuploadAdaptUnifileService.prototype.extendData;
    /** @type {?} */
    FfileuploadAdaptUnifileService.prototype.bufferSize;
    /**
     * @type {?}
     * @private
     */
    FfileuploadAdaptUnifileService.prototype.uploadedChunk;
    /**
     * @type {?}
     * @private
     */
    FfileuploadAdaptUnifileService.prototype.fileTotalChunk;
    /**
     * @type {?}
     * @private
     */
    FfileuploadAdaptUnifileService.prototype.formId;
    /**
     * @type {?}
     * @private
     */
    FfileuploadAdaptUnifileService.prototype.rootId;
    /**
     * @type {?}
     * @private
     */
    FfileuploadAdaptUnifileService.prototype.serverConfig;
    /**
     * @type {?}
     * @private
     */
    FfileuploadAdaptUnifileService.prototype.uploadSer;
    /**
     * @type {?}
     * @private
     */
    FfileuploadAdaptUnifileService.prototype.configSer;
    /**
     * @type {?}
     * @private
     */
    FfileuploadAdaptUnifileService.prototype.inject;
}
