// @farris/extend-fileupload-adapt-unifile/esm2015/lib/ffileupload-adapt-unifile.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ffileupload-adapt-unifile.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FFileUploadModule } from '@farris/extend-file-upload';
import { UploadDialogMoudle } from '@gsp-svc/formdoc-upload';
import { FfileuploadAdaptUnifileService } from './ffileupload-adapt-unifile.service';
import { FfilepreviewAdaptUnifileService } from './ffilepreview-adapt-unifile.service';
import { FFilePreviewAdaptUnifileDirective } from './ffilepreview-adapt-unifile.directive';
import { FileListModule, FileViewerService } from '@gsp-svc/file-viewer';
import { FFileUploadAdaptUnifileConfigToken, FFileUploadAdaptUnifileConfigService } from './ffileupload-adapt-unifile.config';
import { FfilepreviewAdaptSeeimgComponent } from './ffilepreview-adapt-seeimg/ffilepreview-adapt-seeimg.component';
import { FFileAdaptDownloadFileDirective, FFileAdaptPreviewFileDirective } from './ffilepreview-adapt-unifile-simple.directive';
import { LocaleModule } from '@farris/ui-locale';
import { FarrisCommonModule } from '@farris/ui-common';
export class FfileuploadAdaptUnifileModule {
    /**
     * @param {?=} config
     * @return {?}
     */
    static forRoot(config) {
        return {
            ngModule: FfileuploadAdaptUnifileModule,
            providers: [
                { provide: FFileUploadAdaptUnifileConfigToken, useValue: config },
                FFileUploadAdaptUnifileConfigService,
                FfilepreviewAdaptUnifileService,
                FileViewerService
            ]
        };
    }
}
FfileuploadAdaptUnifileModule.decorators = [
    { type: NgModule, args: [{
                declarations: [FFilePreviewAdaptUnifileDirective, FfilepreviewAdaptSeeimgComponent, FFileAdaptDownloadFileDirective, FFileAdaptPreviewFileDirective],
                imports: [
                    CommonModule,
                    UploadDialogMoudle,
                    FileListModule,
                    FFileUploadModule.forRoot(null, FfileuploadAdaptUnifileService),
                    FarrisCommonModule,
                    LocaleModule.forRoot()
                ],
                exports: [
                    FFileUploadModule,
                    FFilePreviewAdaptUnifileDirective,
                    FfilepreviewAdaptSeeimgComponent,
                    FFileAdaptDownloadFileDirective,
                    FFileAdaptPreviewFileDirective
                ],
                providers: [
                    FFileUploadAdaptUnifileConfigService,
                    FfilepreviewAdaptUnifileService,
                    FileViewerService
                ]
            },] }
];
