// @gsp-wf/wf-dimension-def/esm5/lib/entity/dimensiondef.entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var DimensionDefEntity = /** @class */ (function () {
    function DimensionDefEntity() {
    }
    return DimensionDefEntity;
}());
export { DimensionDefEntity };
if (false) {
    /** @type {?} */
    DimensionDefEntity.prototype.id;
    /** @type {?} */
    DimensionDefEntity.prototype.code;
    /** @type {?} */
    DimensionDefEntity.prototype.entryFormID;
    /** @type {?} */
    DimensionDefEntity.prototype.dimensionType;
    /** @type {?} */
    DimensionDefEntity.prototype.dimensionCode;
    /** @type {?} */
    DimensionDefEntity.prototype.dimensionName;
    /** @type {?} */
    DimensionDefEntity.prototype.inputType;
    /** @type {?} */
    DimensionDefEntity.prototype.enumList;
    /** @type {?} */
    DimensionDefEntity.prototype.smartHelp;
    /** @type {?} */
    DimensionDefEntity.prototype.multiSelect;
    /** @type {?} */
    DimensionDefEntity.prototype.helpFilter;
    /** @type {?} */
    DimensionDefEntity.prototype.isUse;
    /** @type {?} */
    DimensionDefEntity.prototype.devLayer;
    /** @type {?} */
    DimensionDefEntity.prototype.contentInfo;
    /** @type {?} */
    DimensionDefEntity.prototype.sortOrder;
    /** @type {?} */
    DimensionDefEntity.prototype.sysInit;
    /** @type {?} */
    DimensionDefEntity.prototype.remarks;
    /** @type {?} */
    DimensionDefEntity.prototype.creator;
    /** @type {?} */
    DimensionDefEntity.prototype.createdTime;
    /** @type {?} */
    DimensionDefEntity.prototype.lastModifier;
    /** @type {?} */
    DimensionDefEntity.prototype.lastModifiedTime;
}
/** @enum {number} */
var DimensionInputType = {
    Help: 0,
    Enum: 1,
};
export { DimensionInputType };
DimensionInputType[DimensionInputType.Help] = 'Help';
DimensionInputType[DimensionInputType.Enum] = 'Enum';
/** @enum {number} */
var DimensionType = {
    BusinessField: 0,
    EnvironmentalVariable: 1,
};
export { DimensionType };
DimensionType[DimensionType.BusinessField] = 'BusinessField';
DimensionType[DimensionType.EnvironmentalVariable] = 'EnvironmentalVariable';
/** @type {?} */
export var WFDimensionTypes = [
    { index: 0, text: '模型字段', value: DimensionType.BusinessField },
];
/** @type {?} */
export var WFDimensionInputTypes = [
    { index: 0, text: '帮助', value: DimensionInputType.Help },
    { index: 1, text: '枚举', value: DimensionInputType.Enum }
];
var ItemEditState = /** @class */ (function () {
    function ItemEditState() {
        this.canEdit = false;
    }
    return ItemEditState;
}());
export { ItemEditState };
if (false) {
    /** @type {?} */
    ItemEditState.prototype.index;
    /** @type {?} */
    ItemEditState.prototype.itemData;
    /** @type {?} */
    ItemEditState.prototype.canEdit;
    /** @type {?} */
    ItemEditState.prototype.id;
}
var EnumItem = /** @class */ (function () {
    function EnumItem() {
    }
    return EnumItem;
}());
export { EnumItem };
if (false) {
    /** @type {?} */
    EnumItem.prototype.key;
    /** @type {?} */
    EnumItem.prototype.value;
}
