// @gsp-wf/wf-dimension-def/esm5/lib/entity/dimendef-entity.util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DimensionType, DimensionInputType } from './dimensiondef.entity';
import { Guid } from './guid.util';
var DimenDefEntityUtil = /** @class */ (function () {
    function DimenDefEntityUtil() {
    }
    /**
     * @param {?} flowFormId
     * @return {?}
     */
    DimenDefEntityUtil.newDimendefEntity = /**
     * @param {?} flowFormId
     * @return {?}
     */
    function (flowFormId) {
        /** @type {?} */
        var newDimendef = {
            id: '',
            code: '',
            entryFormID: '',
            dimensionType: DimensionType.BusinessField,
            dimensionCode: '',
            dimensionName: '',
            inputType: DimensionInputType.Help,
            enumList: null,
            smartHelp: '',
            multiSelect: true,
            helpFilter: '',
            isUse: true,
            devLayer: '',
            contentInfo: null,
            sortOrder: 0,
            sysInit: 0,
            remarks: ''
        };
        newDimendef.id = Guid.newGuid();
        newDimendef.entryFormID = flowFormId;
        return newDimendef;
    };
    return DimenDefEntityUtil;
}());
export { DimenDefEntityUtil };
