// @gsp-aif/runtime-api/esm2015/lib/aif-documents-track/aif-documents-track-manager/aif-documents-track-manager.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output } from '@angular/core';
import { TrackType } from '../config/track-type.map';
import { AifDocumentsTrackService } from '../service/aif-documents-track.service';
import { TrackTypeTextMap } from '../config/track-type-text.map';
/**
 * 单据追踪主组件
 */
export class AifDocumentsTrackManagerComponent {
    /**
     * @param {?} trackService
     */
    constructor(trackService) {
        this.trackService = trackService;
        /**
         * 子组件显示控制
         */
        this.isHidden = true;
        this.isShowable = false;
        this.isResultEmpty = null;
        this.documentId = null;
        this.beTypeCode = null;
        this.selectedTrackType = null;
        this.trackTypes = [];
        /**
         * 传递给子组件的所选择的追踪方式
         */
        this.selectedTrackTypeToChildren = "";
        /**
         * 响应结果转化为组件内传输格式
         */
        this.beTypeCodes = [];
        this.documents = {};
        this.beTypeNames = {};
        //初始化追踪方式下拉框
        for (let item in TrackType) {
            this.trackTypes.push({ value: item, text: TrackTypeTextMap[item] });
        }
        //默认选择“下查”方式
        this.selectedTrackType = this.trackTypes[1].value;
        this.selectedTrackTypeToChildren = this.selectedTrackType;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
    }
    /**
     * @return {?}
     */
    search() {
        this.selectedTrackTypeToChildren = this.selectedTrackType;
        this.trackService.getTrackedDocumentData(this.documentId, this.beTypeCode, this.selectedTrackType).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            /** @type {?} */
            let tempDocuments = {};
            /** @type {?} */
            let tempBeTypeNames = {};
            /** @type {?} */
            let tempBeTypeCodes = [];
            if (this.isNullOrUndefined(data) == false && data.length > 0) {
                this.isResultEmpty = false;
                this.isShowable = true;
                this.isHidden = false;
                for (let item of data) {
                    //查询结果不为空时
                    if (this.isNullOrUndefined(tempDocuments[item.BeTypeCode]) == false) {
                        tempDocuments[item.BeTypeCode].push(item);
                    }
                    else {
                        tempBeTypeCodes.push(item.BeTypeCode);
                        tempBeTypeNames[item.BeTypeCode] = item.BeTypeName;
                        tempDocuments[item.BeTypeCode] = [item];
                    }
                }
            }
            else {
                //查询结果为空时
                this.isResultEmpty = true;
                this.isShowable = false;
                this.isHidden = true;
            }
            this.documents = tempDocuments;
            this.beTypeCodes = tempBeTypeCodes;
            this.beTypeNames = tempBeTypeNames;
        }));
    }
    /**
     * 为外部提供的方便递归调用的查询接口
     * @param {?} documentId
     * @param {?} beType
     * @param {?} trackType
     * @return {?}
     */
    searchFromOut(documentId, beType, trackType) {
        this.documentId = documentId;
        this.beTypeCode = beType;
        this.selectedTrackType = trackType;
        this.search();
    }
    /**
     * 从查询结果中选择单据继续追踪时，父组件接收到子组件传来的
     * 对象，并自动调用查询方法
     * @param {?} $event
     * @return {?}
     */
    trackGoingOn($event) {
        this.documentId = $event.documentId;
        this.beTypeCode = $event.beTypeCode;
        this.selectedTrackType = $event.trackType;
        this.search();
    }
    /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    isNullOrUndefined(arg) {
        return arg == null;
    }
}
AifDocumentsTrackManagerComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-aif-documents-track-manager',
                template: "<div class=\"d-flex g-main\">\r\n  <div class=\"col\">\r\n    <div class=\"g-main d-flex flex-column px-2\">\r\n      <!--\u4E3B\u5185\u5BB9\u533A-->\r\n      <div class=\"mb-1 pr-2 position-relative clearfix\">\r\n        <!-- <div class=\"\" style=\"\"> -->\r\n        <div class=\"position-relative\" [style.display]=\"'inherit'\">\r\n          <div class=\"form-inline row py-1\">\r\n            <div class=\"form-group col-sm-12 col-md-4 mb-1\">\r\n              <div class=\"input-group w-100\">\r\n                <div class=\"input-group-prepend\"><span class=\"input-group-text\">\u5355\u636EID</span></div>\r\n                <input class=\"form-control\" placeholder=\"\" name=\"documentId\" [(ngModel)]=\"documentId\" type=\"text\" />\r\n              </div>\r\n            </div>\r\n\r\n            <div class=\"form-group col-sm-12 col-md-4 mb-1\">\r\n                <div class=\"input-group w-100\">\r\n                  <div class=\"input-group-prepend\"><span class=\"input-group-text\">BE\u7C7B\u578B\u7F16\u53F7</span></div>\r\n                  <input class=\"form-control\" placeholder=\"\" name=\"beTypeCode\" [(ngModel)]=\"beTypeCode\" type=\"text\" />\r\n                </div>\r\n            </div>\r\n\r\n            <div class=\"form-group col-sm-12 col-md-2 mb-1\">\r\n                <div class=\"input-group w-100\">\r\n                  <div class=\"input-group-prepend\"><span class=\"input-group-text\">\u8FFD\u8E2A\u65B9\u5F0F</span></div>\r\n                  <select [(ngModel)]=\"selectedTrackType\" class=\"form-control\">\r\n                      <option *ngFor=\"let tt of trackTypes\" value=\"{{tt.value}}\">{{tt.text}}</option>\r\n                    </select>\r\n                </div>\r\n            </div>\r\n\r\n            <div class=\" col-sm-12 col-md-1 mb-1\">\r\n              <div class=\"w-100\">\r\n                <button class=\"k-button btn btn-sm btn-default\" type=\"button\" (click)=\"search()\">\u67E5\u8BE2</button>\r\n              </div>\r\n            </div>\r\n          </div>\r\n\r\n        </div>\r\n        <div class=\"k-form-field\" style=\"color:red; font-size: unset;\" *ngIf=\"isResultEmpty==true\">\u67E5\u8BE2\u7ED3\u679C\u4E3A\u7A7A</div>\r\n      </div>\r\n\r\n      <div class=\"flex-auto k-form-field pb-1\">\r\n        <app-aif-documents-tree [documents]=\"documents\" [beTypeNames]=\"beTypeNames\" [beTypeCodes]=\"beTypeCodes\"\r\n          [trackType]=\"selectedTrackTypeToChildren\" (trackClick)=\"trackGoingOn($event)\">\r\n        </app-aif-documents-tree>\r\n      </div>\r\n\r\n    </div>\r\n  </div>\r\n</div>",
                styles: [""]
            }] }
];
/** @nocollapse */
AifDocumentsTrackManagerComponent.ctorParameters = () => [
    { type: AifDocumentsTrackService }
];
AifDocumentsTrackManagerComponent.propDecorators = {
    documentId: [{ type: Input }],
    beTypeCode: [{ type: Input }],
    selectedTrackType: [{ type: Input }],
    trackTypes: [{ type: Output }]
};
if (false) {
    /**
     * 子组件显示控制
     * @type {?}
     */
    AifDocumentsTrackManagerComponent.prototype.isHidden;
    /** @type {?} */
    AifDocumentsTrackManagerComponent.prototype.isShowable;
    /** @type {?} */
    AifDocumentsTrackManagerComponent.prototype.isResultEmpty;
    /** @type {?} */
    AifDocumentsTrackManagerComponent.prototype.documentId;
    /** @type {?} */
    AifDocumentsTrackManagerComponent.prototype.beTypeCode;
    /** @type {?} */
    AifDocumentsTrackManagerComponent.prototype.selectedTrackType;
    /** @type {?} */
    AifDocumentsTrackManagerComponent.prototype.trackTypes;
    /**
     * 传递给子组件的所选择的追踪方式
     * @type {?}
     */
    AifDocumentsTrackManagerComponent.prototype.selectedTrackTypeToChildren;
    /**
     * 响应结果转化为组件内传输格式
     * @type {?}
     */
    AifDocumentsTrackManagerComponent.prototype.beTypeCodes;
    /** @type {?} */
    AifDocumentsTrackManagerComponent.prototype.documents;
    /** @type {?} */
    AifDocumentsTrackManagerComponent.prototype.beTypeNames;
    /**
     * @type {?}
     * @private
     */
    AifDocumentsTrackManagerComponent.prototype.trackService;
}
