// @gsp-cmp/ccmp-ui/esm2015/lib/view-utils/view-utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { MessagerService } from '@farris/ui-messager';
import { NotifyService } from '@farris/ui-notify';
import * as i0 from "@angular/core";
import * as i1 from "@farris/ui-messager";
import * as i2 from "@farris/ui-notify";
export class ViewUtils {
    /**
     * @param {?} messagerService
     * @param {?} notify
     */
    constructor(messagerService, notify) {
        this.messagerService = messagerService;
        this.notify = notify;
    }
    /**
     * 获取模态框配置信息
     * @param {?} btnRef 按钮模板引用
     * @param {?=} options 配置项，暂未使用
     * @return {?}
     */
    getModalConfig(btnRef, options) {
        return {
            title: (options && options.title) ? options.title : '操作列表',
            width: (options && options.width) ? options.width : 450,
            height: (options && options.height) ? options.width : 500,
            resizable: (options && options.resizable) ? options.resizable : true,
            showButtons: true,
            showMaxButton: false,
            enableScroll: false,
            buttons: btnRef
        };
    }
    /**
     * @param {?} data
     * @return {?}
     */
    showObseleteDialog(data) {
        this.messagerService.show('warning', `通用构件${data.code}已过时！
                    升级步骤：
                    1. 重新定义通用构件元数据;
                    2. 开发新版本构件代码。
                    请参考GSPCloud开发文档->高级开发->构件开发->通用构件开发`, { safeHtml: false });
    }
    /**
     * @param {?} message
     * @param {?} infoType
     * @param {?} title
     * @param {?} timeout
     * @return {?}
     */
    showInfo(message, infoType, title, timeout) {
        this.notify.config.position = 'top-center';
        if (infoType == 'success') {
            this.notify.success((/** @type {?} */ ({ title: title, msg: message, timeout: timeout })));
        }
        else if (infoType == 'info') {
            this.notify.info((/** @type {?} */ ({ title: title, msg: message, timeout: timeout })));
        }
        else if (infoType == 'warning') {
            this.notify.warning((/** @type {?} */ ({ title: title, msg: message, timeout: timeout })));
        }
        else if (infoType == 'error') {
            this.notify.error((/** @type {?} */ ({ title: title, msg: message, timeout: timeout })));
        }
    }
}
ViewUtils.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ViewUtils.ctorParameters = () => [
    { type: MessagerService },
    { type: NotifyService }
];
/** @nocollapse */ ViewUtils.ngInjectableDef = i0.defineInjectable({ factory: function ViewUtils_Factory() { return new ViewUtils(i0.inject(i1.MessagerService), i0.inject(i2.NotifyService)); }, token: ViewUtils, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ViewUtils.prototype.messagerService;
    /**
     * @type {?}
     * @private
     */
    ViewUtils.prototype.notify;
}
