// @farris/kendo-binding/esm2015/lib/directives/farris-list-filter/farris_list_filter_binding.directive.js
import { Directive, Injector, Input, Optional } from '@angular/core';
import { FrameContext } from '@farris/devkit';
import { ListFilterComponent } from '@farris/ui-list-filter';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
export class FarrisListFilterBindingDirective {
    constructor(injector, frameContext, listFilterComponent) {
        this.injector = injector;
        this.frameContext = frameContext;
        this.listFilterComponent = listFilterComponent;
        /**
         * 变量编号，默认为filterConditionList
         */
        this.variableCode = 'filterConditionList';
        this.$destroy = new Subject();
    }
    ngOnDestroy() {
        if (this.$destroy) {
            this.$destroy.next();
            this.$destroy.complete();
        }
    }
    ngOnInit() {
        if (this.listFilterComponent) {
            this.listFilterComponent.conditionsChange.pipe(takeUntil(this.$destroy)).subscribe((event) => {
                const { data = '[]' } = event || {};
                this.setConditions(data);
            });
        }
    }
    setConditions(conditions) {
        this.frameContext.uiState.setPropertyValue(this.variableCode, conditions);
    }
}
FarrisListFilterBindingDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-list-filter-binding]'
            },] }
];
/** @nocollapse */
FarrisListFilterBindingDirective.ctorParameters = () => [
    { type: Injector },
    { type: FrameContext },
    { type: ListFilterComponent, decorators: [{ type: Optional }] }
];
FarrisListFilterBindingDirective.propDecorators = {
    variableCode: [{ type: Input, args: ['variableCode',] }]
};
