// @gsp-dip/data-imp-exp/esm5/lib/bakcomponents/dataie-progress/dataie-progress.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter } from '@angular/core';
var DataieProgressComponent = /** @class */ (function () {
    function DataieProgressComponent() {
        this.closeModal = new EventEmitter();
    }
    /**
     * @return {?}
     */
    DataieProgressComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    DataieProgressComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-dataie-progress',
                    template: "<form class=\"px-3 px-md-4 farris-form d-flex flex-wrap\">\r\n    <div class=\"col-12\">\r\n\r\n\r\n        <farris-progress [percent]=\"DataIeProgressParams.ImportProgressValue\" [status]=\"'active'\"></farris-progress>\r\n        <p>\u6570\u636E\u5BFC\u5165\u4E2D</p>\r\n\r\n    </div>\r\n</form>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    DataieProgressComponent.ctorParameters = function () { return []; };
    DataieProgressComponent.propDecorators = {
        closeModal: [{ type: Output }]
    };
    return DataieProgressComponent;
}());
export { DataieProgressComponent };
if (false) {
    /** @type {?} */
    DataieProgressComponent.prototype.closeModal;
    /** @type {?} */
    DataieProgressComponent.prototype.DataIeProgressParams;
}
