// @farris/mobile-command-services/esm5/lib/ui-services/dialog.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ui-services/dialog.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { from } from 'rxjs';
import { Dialog } from '@farris/mobile-ui';
var DialogService = /** @class */ (function () {
    function DialogService() {
    }
    /**
     * 提示对话框
     * @summary
     * type=info|success|warning|error
     */
    /**
     * 提示对话框
     * \@summary
     * type=info|success|warning|error
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    DialogService.prototype.alert = /**
     * 提示对话框
     * \@summary
     * type=info|success|warning|error
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    function (message, options) {
        /** @type {?} */
        var promise$ = new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            Dialog.alert(tslib_1.__assign({ title: '提醒', message: message, cancelText: '取消', confirmText: '确定', onConfirm: (/**
                 * @return {?}
                 */
                function () { resolve(true); }), onCancel: (/**
                 * @return {?}
                 */
                function () { resolve(false); }) }, options));
        }));
        return from(promise$);
    };
    /**
     * 确认框对话框
     */
    /**
     * 确认框对话框
     * @param {?} message
     * @return {?}
     */
    DialogService.prototype.confirm = /**
     * 确认框对话框
     * @param {?} message
     * @return {?}
     */
    function (message) {
        /** @type {?} */
        var confirmResult$ = this.confirmPromise(message);
        return from(confirmResult$);
    };
    /**
     * 自定义弹窗
     */
    /**
     * 自定义弹窗
     * @param {?} options
     * @return {?}
     */
    DialogService.prototype.show = /**
     * 自定义弹窗
     * @param {?} options
     * @return {?}
     */
    function (options) {
        this.confirmPromise(options);
    };
    /**
     * 输入对话框
     * @param title 标题
     * @param options 配置信息
     */
    /**
     * 输入对话框
     * @param {?} title 标题
     * @param {?=} options 配置信息
     * @return {?}
     */
    DialogService.prototype.prompt = /**
     * 输入对话框
     * @param {?} title 标题
     * @param {?=} options 配置信息
     * @return {?}
     */
    function (title, options) {
        /** @type {?} */
        var promise$ = new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            /** @type {?} */
            var vm = Dialog.prompt(tslib_1.__assign({ title: title, confirmText: '确定', onConfirm: (/**
                 * @return {?}
                 */
                function () {
                    resolve(vm.$data.promptText);
                }), onCancel: (/**
                 * @return {?}
                 */
                function () { resolve('取消'); }), onShow: (/**
                 * @return {?}
                 */
                function () {
                    console.log('显示弹窗');
                }) }, options));
        }));
        return from(promise$);
    };
    /**
     * 使用Promise实例包装回调方法，以同步回调方法执行后结果
     */
    /**
     * 使用Promise实例包装回调方法，以同步回调方法执行后结果
     * @private
     * @param {?=} message
     * @return {?}
     */
    DialogService.prototype.confirmPromise = /**
     * 使用Promise实例包装回调方法，以同步回调方法执行后结果
     * @private
     * @param {?=} message
     * @return {?}
     */
    function (message) {
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            Dialog.confirm({
                message: message,
                onConfirm: (/**
                 * @return {?}
                 */
                function () { resolve(true); }),
                onCancel: (/**
                 * @return {?}
                 */
                function () { resolve(false); }),
            });
        }));
    };
    return DialogService;
}());
export { DialogService };
