// @farris/mobile-command-services/esm5/lib/ui-services/notify.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ui-services/notify.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Notify } from '@farris/mobile-ui';
/**
 * 消息提示服务
 * \@summary
 * 页面边缘滑出，自动消失，语气略强于Toast
 */
var /**
 * 消息提示服务
 * \@summary
 * 页面边缘滑出，自动消失，语气略强于Toast
 */
NotifyService = /** @class */ (function () {
    function NotifyService() {
    }
    /**
     * 消息提示
     */
    /**
     * 消息提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    NotifyService.prototype.info = /**
     * 消息提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    function (message, options) {
        ((/** @type {?} */ (Notify))).info(message, tslib_1.__assign({ duration: 1500 }, options));
    };
    /**
     * 成功提示
     */
    /**
     * 成功提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    NotifyService.prototype.success = /**
     * 成功提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    function (message, options) {
        ((/** @type {?} */ (Notify))).success(message, tslib_1.__assign({ duration: 1000 }, options));
    };
    /**
     * 警告提示
     */
    /**
     * 警告提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    NotifyService.prototype.warning = /**
     * 警告提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    function (message, options) {
        ((/** @type {?} */ (Notify))).warning(message, tslib_1.__assign({ duration: 1500 }, options));
    };
    /**
     * 失败提示
     */
    /**
     * 失败提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    NotifyService.prototype.error = /**
     * 失败提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    function (message, options) {
        ((/** @type {?} */ (Notify))).error(message, tslib_1.__assign({ duration: 1500 }, options));
    };
    /**
     * 自定义提醒
     */
    /**
     * 自定义提醒
     * @param {?} options
     * @return {?}
     */
    NotifyService.prototype.show = /**
     * 自定义提醒
     * @param {?} options
     * @return {?}
     */
    function (options) {
        ((/** @type {?} */ (Notify))).show(options);
    };
    return NotifyService;
}());
/**
 * 消息提示服务
 * \@summary
 * 页面边缘滑出，自动消失，语气略强于Toast
 */
export { NotifyService };
