// @farris/mobile-command-services/esm5/lib/ui-services/toast.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ui-services/toast.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Toast } from '@farris/mobile-ui';
/**
 * 轻提示服务
 */
var /**
 * 轻提示服务
 */
ToastService = /** @class */ (function () {
    function ToastService() {
    }
    /**
     * 消息提示
     */
    /**
     * 消息提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    ToastService.prototype.info = /**
     * 消息提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    function (message, options) {
        ((/** @type {?} */ (Toast))).info(message, options);
    };
    /**
     * 成功提示
     */
    /**
     * 成功提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    ToastService.prototype.success = /**
     * 成功提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    function (message, options) {
        ((/** @type {?} */ (Toast))).success(message, options);
    };
    /**
     * 警告提示
     */
    /**
     * 警告提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    ToastService.prototype.warning = /**
     * 警告提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    function (message, options) {
        ((/** @type {?} */ (Toast))).warning(message, options);
    };
    /**
     * 失败提示
     */
    /**
     * 失败提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    ToastService.prototype.error = /**
     * 失败提示
     * @param {?} message
     * @param {?=} options
     * @return {?}
     */
    function (message, options) {
        ((/** @type {?} */ (Toast))).error(message, options);
    };
    /**
     * 自定义提示
     */
    /**
     * 自定义提示
     * @param {?=} options
     * @return {?}
     */
    ToastService.prototype.show = /**
     * 自定义提示
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        ((/** @type {?} */ (Toast)))(options);
    };
    return ToastService;
}());
export { ToastService };
