// @farris/mobile-devkit/esm2015/lib/view-model/view_model_context.js
import { Injector, InjectFlags } from '../core/index';
import { AppContext } from '../app/index';
/**
 * 视图模型上下文
 */
class ViewModelContext {
    constructor() {
    }
    init(viewModel) {
        this.innerViewModel = viewModel;
        this.id = this.innerViewModel.id;
        this.appContext = viewModel.injector.get(AppContext);
        this.regToTree();
        this.regToAppContext();
    }
    regToTree() {
        const parentInjector = this.innerViewModel.injector.get(Injector, null, InjectFlags.SkipSelf);
        if (parentInjector) {
            this.parent = parentInjector.get(ViewModelContext, null);
            this.root = this.parent ? this.parent.root : this;
        }
        else {
            this.parent = null;
            this.root = this;
        }
    }
    regToAppContext() {
        this.appContext.regViewModelContext(this);
    }
    get viewModel() {
        return this.innerViewModel;
    }
    get injector() {
        return this.innerViewModel.injector;
    }
    get repository() {
        return this.innerViewModel.repository;
    }
    get bindingData() {
        return this.innerViewModel.bindingData;
    }
    get uiState() {
        return this.innerViewModel.uiState;
    }
    get stateMachine() {
        return this.innerViewModel.stateMachine;
    }
    get form() {
        return this.innerViewModel.form;
    }
    get commandBus() {
        return this.innerViewModel.commandBus;
    }
    get expressionEngineImpl() {
        return this.innerViewModel.expressionEngineImpl;
    }
    get expressionManager() {
        return this.innerViewModel.expressionManager;
    }
    ;
    get expressionResult() {
        return this.innerViewModel.expressionResult;
    }
    ;
    get controlsProxy() {
        return this.innerViewModel.controlsProxy;
    }
    ;
    getVirtualRootFrameContext() {
        let formFrameContext = this;
        // let parent = this.parent;
        // while (parent) {
        //   // 兼容没有重新编译的表单，如果判断parent.namespace存在会导致获取不到root-framecontext
        //   if (parent.namespace === this.namespace) {
        //     formFrameContext = parent;
        //     parent = parent.parent;
        //   } else {
        //     break;
        //   }
        // }
        return formFrameContext;
    }
    /**
     * 注册扩展命令
     */
    registerExtendCommand(commandName, commandHandler) {
        this.viewModel.registerExtendCommand(commandName, commandHandler);
    }
}
export { ViewModelContext };
