// @gsp-cmp/coderule-selector/esm5/lib/service/code-rule-select.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/code-rule-select.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpHeaders, HttpClient } from '@angular/common/http';
var CodeRuleSelectService = /** @class */ (function () {
    function CodeRuleSelectService(http) {
        this.http = http;
        this.QUERY_URL = "/api/runtime/bcc/v1.0/code-rule-data";
        this.metadatas = [];
    }
    /**
     * @param {?} metadata
     * @return {?}
     */
    CodeRuleSelectService.prototype.addMetadataRT = /**
     * @param {?} metadata
     * @return {?}
     */
    function (metadata) {
        this.metadatas.push(metadata);
    };
    /**
     * @return {?}
     */
    CodeRuleSelectService.prototype.clearData = /**
     * @return {?}
     */
    function () {
        this.metadatas = [];
        this.codeRule = null;
    };
    // 根据类型获取Cloud Server下所有的元数据
    // 根据类型获取Cloud Server下所有的元数据
    /**
     * @return {?}
     */
    CodeRuleSelectService.prototype.getCodeRuleList = 
    // 根据类型获取Cloud Server下所有的元数据
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var headers = new HttpHeaders();
        headers.append("sessionId", this.sessionId);
        return this.http.get(this.QUERY_URL + "?serviceUnit=" + this.serviceUnit, { headers: headers });
    };
    CodeRuleSelectService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CodeRuleSelectService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return CodeRuleSelectService;
}());
export { CodeRuleSelectService };
if (false) {
    /** @type {?} */
    CodeRuleSelectService.prototype.serviceUnit;
    /** @type {?} */
    CodeRuleSelectService.prototype.sessionId;
    /**
     * @type {?}
     * @private
     */
    CodeRuleSelectService.prototype.QUERY_URL;
    /** @type {?} */
    CodeRuleSelectService.prototype.codeRule;
    /** @type {?} */
    CodeRuleSelectService.prototype.metadatas;
    /**
     * @type {?}
     * @private
     */
    CodeRuleSelectService.prototype.http;
}
