// @farris/dynamic-control-group/esm5/lib/components/help/help.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { trim } from 'lodash-es';
import { CommonUtils } from '@farris/ui-common';
var HelpComponent = /** @class */ (function () {
    function HelpComponent(utils) {
        var _this = this;
        this.utils = utils;
        /**
         * @param helpData 选中的数据，如果是多选，是数组；如果是单选，是普通对象
         * @param mapFields
         * @param bindingData 帮助所在form的value对象
         */
        this.mapping = (/**
         * @param {?} helpData
         * @param {?} mapFields
         * @return {?}
         */
        function (helpData, mapFields) {
            console.log('helpData ', helpData);
            if (mapFields) {
                /** @type {?} */
                var helpFields = Object.keys(mapFields);
                helpFields.forEach((/**
                 * @param {?} f
                 * @return {?}
                 */
                function (f) {
                    /** @type {?} */
                    var val = '';
                    if (helpData) {
                        if (helpData instanceof Array) {
                            val = helpData.map((/**
                             * @param {?} h
                             * @return {?}
                             */
                            function (h) {
                                return _this.utils.getValue(f, h);
                            })).join(',');
                        }
                        else {
                            val = _this.utils.getValue(f, helpData);
                        }
                    }
                    mapFields[f].split(',').forEach((/**
                     * @param {?} ff
                     * @return {?}
                     */
                    function (ff) {
                        /** @type {?} */
                        var field = trim(ff);
                        // this.utils.setValue(bindingData, field, val);
                        _this.form.controls[field].setValue(val);
                    }));
                }));
            }
        });
    }
    /**
     * @return {?}
     */
    HelpComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.options = this.control.control;
        this.className = this.control.className ? this.control.className : 'col-12 col-md-6 col-xl-3 col-el-2';
    };
    HelpComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-help",
                    template: "\n    <div class=\"farris-group-wrap\" [formGroup]=\"form\">\n      <div class=\"form-group farris-form-group\">\n        <label class=\"col-form-label\">\n          <span *ngIf=\"0\" class=\"farris-label-info text-danger\">*</span>\n          <span class=\"farris-label-text\">{{control.name}}</span>\n        </label>\n        <div class=\"farris-input-wrap\">\n            <farris-lookup-grid\n              [formControlName]=\"control.labelCode\"\n              [uri]=\"options.uri\"\n              [singleSelect]=\"options.singleSelect\"\n              [displayType]=\"options.displayType\"\n              [idField]=\"options.idField\"\n              [textField]=\"options.textField\"\n              [valueField]=\"options.valueField\"\n              [title]=\"options.dialogTitle\"\n              [mappingFn]=\"mapping\"\n              [mapFields]=\"options.mapFields\"\n              [readonly]=\"options.readonly\"\n          ></farris-lookup-grid>\n        </div>\n      </div>\n    </div>\n  "
                }] }
    ];
    /** @nocollapse */
    HelpComponent.ctorParameters = function () { return [
        { type: CommonUtils }
    ]; };
    HelpComponent.propDecorators = {
        className: [{ type: HostBinding, args: ['class',] }],
        control: [{ type: Input }],
        form: [{ type: Input }]
    };
    return HelpComponent;
}());
export { HelpComponent };
if (false) {
    /** @type {?} */
    HelpComponent.prototype.className;
    /** @type {?} */
    HelpComponent.prototype.control;
    /** @type {?} */
    HelpComponent.prototype.form;
    /** @type {?} */
    HelpComponent.prototype.options;
    /**
     * \@param helpData 选中的数据，如果是多选，是数组；如果是单选，是普通对象
     * \@param mapFields
     * \@param bindingData 帮助所在form的value对象
     * @type {?}
     */
    HelpComponent.prototype.mapping;
    /**
     * @type {?}
     * @private
     */
    HelpComponent.prototype.utils;
}
