// @farris/ui-datagrid/esm2015/lib/types/data-column.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function CustomStyle() { }
if (false) {
    /** @type {?|undefined} */
    CustomStyle.prototype.cls;
    /** @type {?|undefined} */
    CustomStyle.prototype.style;
}
/**
 * @record
 */
export function CellContext() { }
if (false) {
    /** @type {?} */
    CellContext.prototype.field;
    /** @type {?} */
    CellContext.prototype.rowIndex;
    /** @type {?} */
    CellContext.prototype.rowData;
    /** @type {?} */
    CellContext.prototype.column;
    /** @type {?} */
    CellContext.prototype.instance;
    /** @type {?} */
    CellContext.prototype.formGroup;
    /** @type {?} */
    CellContext.prototype.tools;
}
/**
 * @record
 */
export function ICellTemplateComponent() { }
if (false) {
    /** @type {?} */
    ICellTemplateComponent.prototype.ctx;
    /** @type {?|undefined} */
    ICellTemplateComponent.prototype.bindingContext;
}
/**
 * @record
 */
export function DataColumn() { }
if (false) {
    /** @type {?|undefined} */
    DataColumn.prototype.dataType;
    /** @type {?|undefined} */
    DataColumn.prototype.id;
    /** @type {?|undefined} */
    DataColumn.prototype.parentId;
    /** @type {?} */
    DataColumn.prototype.field;
    /** @type {?} */
    DataColumn.prototype.title;
    /** @type {?} */
    DataColumn.prototype.width;
    /**
     * 记录原始定义宽度
     * @type {?|undefined}
     */
    DataColumn.prototype.originalWidth;
    /**
     * 标题对齐方式
     * @type {?|undefined}
     */
    DataColumn.prototype.halign;
    /**
     * 文本对齐方式
     * @type {?|undefined}
     */
    DataColumn.prototype.align;
    /**
     * 垂直对齐方式
     * @type {?|undefined}
     */
    DataColumn.prototype.valign;
    /** @type {?|undefined} */
    DataColumn.prototype.formatter;
    /** @type {?|undefined} */
    DataColumn.prototype.styler;
    /**
     * 表头自定义样式
     * @type {?|undefined}
     */
    DataColumn.prototype.headerStyler;
    /**
     * 左侧边距
     * @type {?|undefined}
     */
    DataColumn.prototype.left;
    /**
     * 右侧边距
     * @type {?|undefined}
     */
    DataColumn.prototype.right;
    /**
     * 是否固定
     * @type {?|undefined}
     */
    DataColumn.prototype.fixed;
    /**
     * 是否显示
     * @type {?|undefined}
     */
    DataColumn.prototype.visible;
    /**
     * 只读
     * @type {?|undefined}
     */
    DataColumn.prototype.readonly;
    /** @type {?|undefined} */
    DataColumn.prototype.editor;
    /**
     * 单元格自定义模板
     * @type {?|undefined}
     */
    DataColumn.prototype.template;
    /**
     * 表关模板
     * @type {?|undefined}
     */
    DataColumn.prototype.headerTemplate;
    /**
     * 编辑器模板
     * @type {?|undefined}
     */
    DataColumn.prototype.editorTemplate;
    /**
     * 鼠标移动至单元格后，显示悬浮消息
     * @type {?|undefined}
     */
    DataColumn.prototype.showTips;
    /**
     * 单元格提示模式：
     *  allways： 鼠标滑过即显示
     *  auto: 单元格宽度不够时才会显示
     * @type {?|undefined}
     */
    DataColumn.prototype.tipMode;
    /** @type {?|undefined} */
    DataColumn.prototype.tipContent;
    /** @type {?|undefined} */
    DataColumn.prototype.tipOptions;
    /**
     * True to allow the column can be sorted.
     * @type {?|undefined}
     */
    DataColumn.prototype.sortable;
    /** @type {?|undefined} */
    DataColumn.prototype.order;
    /** @type {?|undefined} */
    DataColumn.prototype.sorter;
    /**
     * True to allow the column can be resized.
     * @type {?|undefined}
     */
    DataColumn.prototype.resizable;
    /** @type {?|undefined} */
    DataColumn.prototype.rowspan;
    /**
     * 列合并
     * @type {?|undefined}
     */
    DataColumn.prototype.colspan;
    /**
     * 列合并原始值
     * @type {?|undefined}
     */
    DataColumn.prototype.origianlColSpan;
    /** @type {?|undefined} */
    DataColumn.prototype.index;
    /** @type {?|undefined} */
    DataColumn.prototype.footer;
    /**
     * 允许分组，默认为 true
     * @type {?|undefined}
     */
    DataColumn.prototype.allowGrouping;
    /** @type {?|undefined} */
    DataColumn.prototype.groupFooter;
    /** @type {?|undefined} */
    DataColumn.prototype.validators;
    /** @type {?|undefined} */
    DataColumn.prototype.filter;
    /**
     * 是否多语字段
     * @type {?|undefined}
     */
    DataColumn.prototype.isMultilingualField;
    /**
     * 空格键响应回调函数
     * @type {?|undefined}
     */
    DataColumn.prototype.onSpaceHandler;
    /**
     * 回车响应回调函数
     * @type {?|undefined}
     */
    DataColumn.prototype.onEnterHandler;
    /** @type {?|undefined} */
    DataColumn.prototype.componentType;
    /** @type {?|undefined} */
    DataColumn.prototype.headerComponentType;
    /**
     * 由datagrid接管click 事件，默认为 true
     * @type {?|undefined}
     */
    DataColumn.prototype.proxyClick;
}
/** @enum {number} */
const ColumnFilterType = {
    string: 0,
    number: 1,
    boolean: 2,
    date: 3,
    time: 4,
    datetime: 5,
    enum: 6,
    fromdata: 7,
};
export { ColumnFilterType };
ColumnFilterType[ColumnFilterType.string] = 'string';
ColumnFilterType[ColumnFilterType.number] = 'number';
ColumnFilterType[ColumnFilterType.boolean] = 'boolean';
ColumnFilterType[ColumnFilterType.date] = 'date';
ColumnFilterType[ColumnFilterType.time] = 'time';
ColumnFilterType[ColumnFilterType.datetime] = 'datetime';
ColumnFilterType[ColumnFilterType.enum] = 'enum';
ColumnFilterType[ColumnFilterType.fromdata] = 'fromdata';
/**
 * Client Filter Type
 * @record
 */
export function FilterCondition() { }
if (false) {
    /** @type {?|undefined} */
    FilterCondition.prototype.operator1;
    /** @type {?|undefined} */
    FilterCondition.prototype.value1;
    /** @type {?|undefined} */
    FilterCondition.prototype.relation;
    /** @type {?|undefined} */
    FilterCondition.prototype.operator2;
    /** @type {?|undefined} */
    FilterCondition.prototype.value2;
}
/**
 * @record
 */
export function ColumnFilter() { }
if (false) {
    /** @type {?|undefined} */
    ColumnFilter.prototype.type;
    /** @type {?|undefined} */
    ColumnFilter.prototype.enableOperator;
    /** @type {?|undefined} */
    ColumnFilter.prototype.options;
}
/**
 * @record
 */
export function GridEditor() { }
if (false) {
    /** @type {?} */
    GridEditor.prototype.type;
    /** @type {?|undefined} */
    GridEditor.prototype.options;
    /** @type {?|undefined} */
    GridEditor.prototype.validators;
}
/**
 * @record
 */
export function ColumnGroup() { }
if (false) {
    /**
     * 左侧固定列总宽度
     * @type {?|undefined}
     */
    ColumnGroup.prototype.leftFixedWidth;
    /**
     * 左侧固定列集合
     * @type {?|undefined}
     */
    ColumnGroup.prototype.leftFixed;
    /**
     * 非固定列集合
     * @type {?|undefined}
     */
    ColumnGroup.prototype.normalColumns;
    /**
     * 右侧固定列集合
     * @type {?|undefined}
     */
    ColumnGroup.prototype.rightFixed;
    /** @type {?|undefined} */
    ColumnGroup.prototype.columns;
    /**
     * 右侧固定列宽度
     * @type {?|undefined}
     */
    ColumnGroup.prototype.rightFixedWidth;
    /**
     * 正常宽度
     * @type {?|undefined}
     */
    ColumnGroup.prototype.normalWidth;
    /**
     * 所有列宽度之和
     * @type {?|undefined}
     */
    ColumnGroup.prototype.totalWidth;
    /**
     * 用户设置的列宽之和
     * @type {?|undefined}
     */
    ColumnGroup.prototype.originalTotalWidth;
}
/**
 * @record
 */
export function PaginationInfo() { }
if (false) {
    /** @type {?|undefined} */
    PaginationInfo.prototype.enable;
    /** @type {?|undefined} */
    PaginationInfo.prototype.pageList;
    /** @type {?|undefined} */
    PaginationInfo.prototype.pageSize;
    /** @type {?|undefined} */
    PaginationInfo.prototype.pageIndex;
    /** @type {?|undefined} */
    PaginationInfo.prototype.total;
}
/** @enum {number} */
const CalculationType = {
    /** 无 */
    none: -1,
    /** 最大值 */
    max: 0,
    /** 最小值 */
    min: 1,
    /** 求和 */
    sum: 2,
    /** 平均数 */
    average: 3,
    /** 总记录数 */
    count: 4,
};
export { CalculationType };
CalculationType[CalculationType.none] = 'none';
CalculationType[CalculationType.max] = 'max';
CalculationType[CalculationType.min] = 'min';
CalculationType[CalculationType.sum] = 'sum';
CalculationType[CalculationType.average] = 'average';
CalculationType[CalculationType.count] = 'count';
/**
 * @record
 */
export function GridFooterRow() { }
if (false) {
    /** @type {?|undefined} */
    GridFooterRow.prototype.options;
    /**
     * 文本对齐方式
     * @type {?|undefined}
     */
    GridFooterRow.prototype.align;
    /**
     * 数据格式化
     * @type {?|undefined}
     */
    GridFooterRow.prototype.formatter;
    /**
     * 单元格样式
     * @type {?|undefined}
     */
    GridFooterRow.prototype.styler;
    /** @type {?|undefined} */
    GridFooterRow.prototype.rowspan;
    /** @type {?|undefined} */
    GridFooterRow.prototype.colspan;
    /** @type {?|undefined} */
    GridFooterRow.prototype.visible;
}
/** @type {?} */
export const defaultPaginationInfo = {
    enable: true,
    pageIndex: 1,
    pageSize: 10
};
