// @farris/mobile-command-services/esm5/lib/map-services/static-map.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/map-services/static-map.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var StaticMap = /** @class */ (function () {
    function StaticMap() {
        // 地图尺寸
        this.size = { width: 400, height: 400 };
        // 标记尺寸  三种
        this.markerSize = 'mid';
        // 标记数组
        this.markers = [];
    }
    return StaticMap;
}());
export { StaticMap };
if (false) {
    /** @type {?} */
    StaticMap.prototype.key;
    /** @type {?} */
    StaticMap.prototype.location;
    /** @type {?} */
    StaticMap.prototype.zoom;
    /** @type {?} */
    StaticMap.prototype.size;
    /** @type {?} */
    StaticMap.prototype.scale;
    /** @type {?} */
    StaticMap.prototype.markerColor;
    /** @type {?} */
    StaticMap.prototype.markerSize;
    /** @type {?} */
    StaticMap.prototype.markers;
    /** @type {?} */
    StaticMap.prototype.labels;
    /** @type {?} */
    StaticMap.prototype.paths;
    /** @type {?} */
    StaticMap.prototype.traffic;
    /** @type {?} */
    StaticMap.prototype.sig;
}
var MapLocation = /** @class */ (function () {
    function MapLocation() {
    }
    return MapLocation;
}());
if (false) {
    /** @type {?} */
    MapLocation.prototype.longtitude;
    /** @type {?} */
    MapLocation.prototype.latitude;
}
var MapSize = /** @class */ (function () {
    function MapSize() {
    }
    return MapSize;
}());
if (false) {
    /** @type {?} */
    MapSize.prototype.width;
    /** @type {?} */
    MapSize.prototype.height;
}
var MapMarker = /** @class */ (function () {
    function MapMarker() {
    }
    return MapMarker;
}());
if (false) {
    /** @type {?} */
    MapMarker.prototype.longtitude;
    /** @type {?} */
    MapMarker.prototype.latitude;
}
