// @farris/kendo-binding/esm2015/lib/directives/farris_checkbox_modification.directive.js
import { Directive, ElementRef, Injector, Input } from '@angular/core';
export class FarrisCheckboxModificationDirective {
    constructor(injector, elementRef) {
        this.injector = injector;
        this.elementRef = elementRef;
    }
    ngOnInit() {
        if (typeof this.index !== 'number' || !this.elementRef) {
            return;
        }
        const nativeElement = this.elementRef.nativeElement;
        if (nativeElement) {
            const id = nativeElement.getAttribute('id');
            if (id) {
                const token = `${id}_${this.index}`;
                this.elementRef.nativeElement.setAttribute('id', token);
                this.elementRef.nativeElement.setAttribute('name', token);
                this.elementRef.nativeElement.setAttribute('original_id', id);
            }
        }
    }
}
FarrisCheckboxModificationDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-checkbox-modification]'
            },] }
];
/** @nocollapse */
FarrisCheckboxModificationDirective.ctorParameters = () => [
    { type: Injector },
    { type: ElementRef }
];
FarrisCheckboxModificationDirective.propDecorators = {
    index: [{ type: Input }]
};
