// @farris/kendo-binding/esm2015/lib/directives/farris-sort.directive.js
import { Directive, Input } from '@angular/core';
export class FarrisSortDirective {
}
FarrisSortDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farrisSort]'
            },] }
];
export class FarrisTextareaEditDirective {
}
FarrisTextareaEditDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-textarea-edit]'
            },] }
];
export class FarrisDynamicColumnFormatDirective {
}
FarrisDynamicColumnFormatDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farrisDynamicColumnFormat]'
            },] }
];
export class FarrisEditEnterDirective {
}
FarrisEditEnterDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-enter-edit]'
            },] }
];
FarrisEditEnterDirective.propDecorators = {
    farrisForm: [{ type: Input, args: ['form-group',] }]
};
export class FarrisTreeTableFormatDirective {
}
FarrisTreeTableFormatDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farrisTreeTableFormat]'
            },] }
];
