// @farris/kendo-binding/esm2015/lib/directives/farris-set-focus.directive.js
import { Directive, Input, Renderer2, Injector, ElementRef } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
export class FarrisSetFocusDirective {
    constructor(rd, injector, el) {
        this.rd = rd;
        this.injector = injector;
        this.el = el;
        this._state = new BehaviorSubject({});
    }
    set focusState(value) {
        this._state.next(value);
    }
    ngOnInit() {
        this.selfEl = this.el.nativeElement;
        this._state.subscribe(data => {
            if (data === false && this.selfEl) {
                if (this.selfEl.nodeName.toLowerCase() === 'farris-html-editor') {
                    this.setHtmlEditorFocus();
                }
                else {
                    this.setFocus();
                }
            }
        });
    }
    /**
     * 设置一般输入框的默认焦点
     */
    setFocus() {
        const input = this.selfEl.querySelectorAll('input');
        if (input.length) {
            input[0].focus();
        }
        else {
            this.selfEl.focus();
        }
    }
    /**
     * 设置html编辑器的默认焦点  总是在文本最后
     */
    setHtmlEditorFocus() {
        // 富文本编辑器不是input
        const ediotrArea = this.selfEl.querySelector('div[contenteditable="true"]');
        //创建range  存在浏览器兼容性问题
        const range = window.getSelection();
        //range 选择obj下所有子内容
        range.selectAllChildren(ediotrArea);
        //光标移至最后
        range.collapseToEnd();
    }
}
FarrisSetFocusDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-set-focus]'
            },] }
];
/** @nocollapse */
FarrisSetFocusDirective.ctorParameters = () => [
    { type: Renderer2 },
    { type: Injector },
    { type: ElementRef }
];
FarrisSetFocusDirective.propDecorators = {
    focusState: [{ type: Input }]
};
