// @farris/kendo-binding/esm2015/lib/directives/farris_editor.directive.js
import { Directive, ElementRef } from '@angular/core';
import { ViewModel } from '@farris/devkit';
import { EditorComponent } from '@farris/ui-editor';
export class FarrisEditorDirective {
    constructor(viewModel, editor) {
        this.viewModel = viewModel;
        this.editor = editor;
    }
    ngOnInit() {
        this.setComponentRef();
    }
    setComponentRef() {
        const appContext = this.viewModel && this.viewModel.frameContext && this.viewModel.frameContext.getFormAppContext();
        const frameId = this.viewModel && this.viewModel.frameContext && this.viewModel.frameContext.frameId;
        const editor = this.editor;
        const elementRef = editor && editor.injector && editor.injector.get(ElementRef, null);
        const id = elementRef && elementRef.nativeElement && elementRef.nativeElement.id || null;
        // 如果frameId不存在或treetable没有id属性，说明不符合使用场景
        if (!frameId || !id) {
            return;
        }
        const map = appContext && appContext.componentRefs && appContext.componentRefs.get(frameId) || new Map();
        appContext && appContext.componentRefs && appContext.componentRefs.set(frameId, map.set(id, this.editor));
    }
}
FarrisEditorDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-editor],[farris-editor-bind]'
            },] }
];
/** @nocollapse */
FarrisEditorDirective.ctorParameters = () => [
    { type: ViewModel },
    { type: EditorComponent }
];
