// @farris/kendo-binding/esm2015/lib/directives/farris_querysolution_event_bind.directive.js
import { Directive, HostListener, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { QuerySolutionComponent } from "@gsp-cmp/querysolution";
export class FarrisQuerySolutionEventBindDirective {
    constructor(injector, component, frameContext) {
        this.injector = injector;
        this.component = component;
        this.frameContext = frameContext;
    }
    /**
     * 执行筛选
     * @param event event
     */
    onQuery(event) {
        if (this.frameContext) {
            this.frameContext.repository.entityCollection.pageIndex = 1;
        }
    }
}
FarrisQuerySolutionEventBindDirective.decorators = [
    { type: Directive, args: [{
                selector: "[farris-querysolution-event-bind]"
            },] }
];
/** @nocollapse */
FarrisQuerySolutionEventBindDirective.ctorParameters = () => [
    { type: Injector },
    { type: QuerySolutionComponent },
    { type: FrameContext }
];
FarrisQuerySolutionEventBindDirective.propDecorators = {
    onQuery: [{ type: HostListener, args: ["queryEmitter",] }]
};
