// @farris/kendo-binding/esm2015/lib/directives/farris_label_modification.directive.js
import { Directive, ElementRef, Injector, Input } from '@angular/core';
export class FarrisLabelModificationDirective {
    constructor(injector, elementRef) {
        this.injector = injector;
        this.elementRef = elementRef;
    }
    ngOnInit() {
        if (typeof this.index !== 'number' || !this.elementRef) {
            return;
        }
        const nativeElement = this.elementRef.nativeElement;
        if (nativeElement) {
            const labelFor = nativeElement.getAttribute('for');
            if (labelFor) {
                this.elementRef.nativeElement.setAttribute('for', `${labelFor}_${this.index}`);
            }
        }
    }
}
FarrisLabelModificationDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-label-modification]'
            },] }
];
/** @nocollapse */
FarrisLabelModificationDirective.ctorParameters = () => [
    { type: Injector },
    { type: ElementRef }
];
FarrisLabelModificationDirective.propDecorators = {
    index: [{ type: Input }]
};
