// @farris/kendo-binding/esm2015/lib/directives/grid-summary.directive.js
import { Directive, Input, NgZone } from '@angular/core';
import { BindingData } from '@farris/devkit';
import { ColumnComponent } from '@progress/kendo-angular-grid';
export class GridColumnAggregateDirective {
    constructor(column, ngZone) {
        this.column = column;
        this.ngZone = ngZone;
    }
    get field() {
        return this.column.field;
    }
    get aggrType() {
        let type = 'none';
        if (typeof (this.aggregate) === 'boolean') {
            // 默认为sum，如果使用指令时传入的input为布尔类型，使用sum。
            if (this.aggregate) {
                type = 'sum';
            }
        }
        else if (typeof (this.aggregate) === 'string') {
            type = this.aggregate;
        }
        return type;
    }
    recalculateSummary(bindingList) {
        let summary = null;
        if (this.aggrType === 'sum') {
            const plainData = bindingList.toJSON().map(item => {
                // 考虑帮助和udt的情况，field是a.b的格式。
                return this.field.split('.').reduce((prevPath, currPath) => {
                    return prevPath[currPath];
                }, item);
            });
            if (plainData && plainData.length) {
                summary = plainData.reduce((prev, curr) => {
                    return prev + curr;
                });
            }
        }
        else if (this.aggrType === 'count') {
            // 计数，如果启用分页，只计算当前页，因此使用length而非total。
            summary = bindingList.length;
        }
        this.updateFooter(summary);
    }
    updateFooter(content) {
        if (this.aggregate) {
            // 存在性能问题，去掉setTimeout。
            // 列表的列越多，此处问题越严重，需要正向解决以下错误。
            // todo: settimeout是为了解决数据变化时angular报错，如果更新页脚的逻辑不在gridbinding发起，是否还存在问题？
            // setTimeout(() => {
            this.column['summary'] = content;
            // });
        }
    }
}
GridColumnAggregateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[aggregate]',
            },] }
];
/** @nocollapse */
GridColumnAggregateDirective.ctorParameters = () => [
    { type: ColumnComponent },
    { type: NgZone }
];
GridColumnAggregateDirective.propDecorators = {
    aggregate: [{ type: Input }]
};
export class GridSummaryDirective /* implements OnInit, AfterViewInit*/ {
    constructor(bindingData) {
        this.bindingData = bindingData;
    }
}
GridSummaryDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farrisSummary]'
            },] }
];
/** @nocollapse */
GridSummaryDirective.ctorParameters = () => [
    { type: BindingData }
];
GridSummaryDirective.propDecorators = {
    summary: [{ type: Input, args: ['farrisSummary',] }]
};
