// @farris/kendo-binding/esm2015/lib/directives/lookup-enable-extend-load-method.directive.js
import { Directive, Optional, Self, Input } from '@angular/core';
import { FrameContext } from '@farris/devkit';
import { LookupGridComponent } from '@farris/ui-lookup';
// tslint:disable-next-line: directive-selector
export class LookupEnableExtendLoadMethodDirective {
    constructor(frameContext, lookup) {
        this.frameContext = frameContext;
        this.lookup = lookup;
    }
    ngOnInit() {
        if (!this.frameContext || !this.lookup || this.enableExtendLoadMethod !== true) {
            return;
        }
        const befApiUrl = this.frameContext.repository.apiUri;
        const helpMetadataId = this.lookup.uri;
        const enableKey = `${helpMetadataId}@${befApiUrl}`;
        this.frameContext.setParam(enableKey, true);
    }
}
LookupEnableExtendLoadMethodDirective.decorators = [
    { type: Directive, args: [{
                selector: '[enableExtendLoadMethod]'
            },] }
];
/** @nocollapse */
LookupEnableExtendLoadMethodDirective.ctorParameters = () => [
    { type: FrameContext, decorators: [{ type: Optional }] },
    { type: LookupGridComponent, decorators: [{ type: Optional }, { type: Self }] }
];
LookupEnableExtendLoadMethodDirective.propDecorators = {
    enableExtendLoadMethod: [{ type: Input }]
};
