// @farris/kendo-binding/esm2015/lib/directives/lookup-data-mapping.directive.js
/*
 * @Author: 疯狂秀才(lucas huang)
 * @Date: 2018-11-07 16:31:57
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-09-27 14:57:22
 * @Company: Inspur
 * @Version: v0.0.1
 */
/**
 * 使用方法：
 * [data-mapping]="{ id: 'user.userId', name: 'user.userName' }"
 * key 为帮助上的字段， value 为 表单中的字段名
 * 帮助上的同一个字段可以映射到表单中的多个字段中，{ ... id: 'user.userid, user.addusid'}
 * 多字段以逗号隔开
 *
 */
import { Directive, Optional, Self, Input } from '@angular/core';
import { BindingObject, ViewModel } from '@farris/devkit';
import { LookupGridComponent } from '@farris/ui-lookup';
import { DataMapping } from './data-mapping';
export class LookupDataMappingDirective extends DataMapping {
    constructor(vm, lookup) {
        super();
        this.vm = vm;
        this.lookup = lookup;
        this.target = null;
    }
    ngOnInit() {
        this.lookup.selectedData.subscribe((data) => {
            const _mapfields = this.mapfields || this.lookup.mapFields;
            const spliter = this.lookup.multipleChoiceSeparator || this.defaultSpliter;
            this.mappingData(data, _mapfields, false, spliter);
        });
        this.lookup.clearMappings.subscribe((result) => {
            // const value = result && result.value || null;
            const mapfields = Object.assign({}, (this.mapfields || this.lookup.mapFields || {}));
            const lookupTextField = this.lookup.textField;
            const data = {};
            const controlName = this.lookup.ngControl && this.lookup.ngControl.name;
            if (controlName && this.vm) {
                const textFieldMapping = mapfields[lookupTextField];
                const ngFormControl = this.vm && this.vm.form && this.vm.form.ngFormControls && this.vm.form.ngFormControls[controlName];
                const binding = ngFormControl && ngFormControl.binding;
                if (textFieldMapping && binding) {
                    mapfields[lookupTextField] = textFieldMapping.split(',').filter((item) => item !== binding).join(',');
                }
            }
            // this.setValue(data, lookupTextField.split('.'), value);
            Object.keys(mapfields).forEach((field) => {
                this.setValue(data, field.split('.'), '');
            });
            this.mappingData(data, mapfields, true);
        });
        this.lookup.clear.subscribe(() => {
            const _mapfields = this.mapfields || this.lookup.mapFields;
            this.mappingData(null, _mapfields);
        });
    }
}
LookupDataMappingDirective.decorators = [
    { type: Directive, args: [{ selector: '[data-mapping]' },] }
];
/** @nocollapse */
LookupDataMappingDirective.ctorParameters = () => [
    { type: ViewModel, decorators: [{ type: Optional }] },
    { type: LookupGridComponent, decorators: [{ type: Optional }, { type: Self }] }
];
LookupDataMappingDirective.propDecorators = {
    mapfields: [{ type: Input, args: ['data-mapping',] }],
    target: [{ type: Input, args: ['target',] }]
};
