// @farris/kendo-binding/esm2015/lib/directives/uistate-binding.directive.js
import { Directive, Input, HostListener, Self, Host, Optional, Output, EventEmitter, KeyValueDiffers } from '@angular/core';
import { ComboBoxComponent } from '@progress/kendo-angular-dropdowns';
import { LookupGridComponent } from '@farris/ui-lookup';
import { ComboListComponent } from '@farris/ui-combo-list';
import { ComboLookupComponent } from '@farris/ui-combo-lookup';
export class UIStateBindingDirective {
    constructor(hostComboComponent, hostHelpComponent, differs, hostComboListComponent, hostComboLookupComponent) {
        this.hostComboComponent = hostComboComponent;
        this.hostHelpComponent = hostHelpComponent;
        this.differs = differs;
        this.hostComboListComponent = hostComboListComponent;
        this.hostComboLookupComponent = hostComboLookupComponent;
        this.defaultSpliter = ',';
        this.differ = null;
        this.UIStateBindingChange = new EventEmitter();
    }
    set bindingObject(value) {
        this._bindingObject = value;
        if (!this.differ && value && this.differs && typeof (value) === 'object') {
            this.differ = this.differs.find(value).create();
        }
    }
    get bindingObject() {
        return this._bindingObject;
    }
    onValueChange(val) {
        // host is combobox
        if (this.hostComboComponent && this.bindingObject) {
            if (val) {
                this.bindingObject.key = val.value;
                this.bindingObject.value = val.name;
            }
            else {
                this.bindingObject.key = null;
                this.bindingObject.value = null;
            }
        }
    }
    ngOnInit() {
        if (this.hostComboComponent) {
            this.hostComboComponent.valuePrimitive = false;
        }
        else if (this.hostHelpComponent) {
            this.bindObjectToHostLookup();
        }
        else if (this.hostComboListComponent) {
            this.bindObjectToHostComboList();
        }
        else if (this.hostComboLookupComponent) {
            this.bindObjectToHostComboLookup();
        }
    }
    ngDoCheck() {
        if (this.differ && typeof (this.bindingObject) === 'object') {
            const changes = this.differ && this.differ.diff(this.bindingObject);
            if (changes) {
                this.bindingChanges();
            }
        }
        else { // 兼容未重新编译工程，differ不存在从情况
            this.bindingChanges();
        }
    }
    bindingChanges() {
        const text = this.bindingObject ? this.bindingObject.value : null;
        const key = this.bindingObject ? this.bindingObject.key : null;
        if (this.hostComboComponent) {
            this.hostComboComponent.text = text;
            const vField = this.hostComboComponent.valueField;
            const item = this.hostComboComponent.data.find((n) => n[vField] === key);
            this.hostComboComponent.writeValue(item);
        }
        else if (this.hostHelpComponent) {
            this.hostHelpComponent.writeValue(text);
            this.hostHelpComponent.displayValue = key;
        }
        else if (this.hostComboListComponent) {
            this.hostComboListComponent.writeValue(key);
        }
        else if (this.hostComboLookupComponent) {
            this.hostComboLookupComponent.selectedValues = key;
            this.hostComboLookupComponent.writeValue(text);
        }
    }
    ngOnChanges(changes) {
        if (changes.bindingObject && !this.differ) {
            this.bindingChanges();
        }
    }
    // 弹出帮助
    bindObjectToHostLookup() {
        if (!this.hostHelpComponent) {
            return;
        }
        this.hostHelpComponent.selectedData.subscribe((data) => this.updateHelpBindingObject(data));
        this.hostHelpComponent.clear.subscribe(() => {
            // this.bindingObject = {key: null, value: null};
            this.updateHelpBindingObject(null);
        });
        if (this.hostHelpComponent.tagRemoved) {
            this.hostHelpComponent.tagRemoved.subscribe((event) => {
                const { removedIndex: index = -1, instance: componentRef = null } = event;
                if (index === -1) {
                    return;
                }
                const isSigleSelect = componentRef.singleSelect;
                if (isSigleSelect || !componentRef.displayValue) {
                    this.updateHelpBindingObject(null);
                }
                else {
                    this.UIStateBindingChange.emit({ key: componentRef.displayValue, value: componentRef.displayText });
                }
            });
        }
        if (this.hostHelpComponent.nosearch) {
            this.hostHelpComponent.valueChanged.subscribe((txt) => {
                const idField = this.hostHelpComponent.idField;
                const textField = this.hostHelpComponent.textField;
                this.updateHelpBindingObject({
                    [idField]: txt,
                    [textField]: txt
                });
            });
        }
    }
    // 下拉列表
    bindObjectToHostComboList() {
        if (!this.hostComboListComponent) {
            return;
        }
        this.hostComboListComponent.valueChange.subscribe((data) => this.updateHelpBindingObject(data.selections));
        this.hostComboListComponent.clear.subscribe(() => {
            this.updateHelpBindingObject(null);
        });
    }
    // 下拉帮助
    bindObjectToHostComboLookup() {
        if (!this.hostComboLookupComponent) {
            return;
        }
        if (this.hostComboLookupComponent.multiSelect) {
            this.hostComboLookupComponent.valueChange.subscribe((data) => this.updateHelpBindingObject(data.selections));
        }
        else {
            // this.hostComboLookupComponent.selectChange.subscribe((data: any) => this.updateHelpBindingObject(data));
            this.hostComboLookupComponent.selectChange.subscribe((e) => {
                let data = e;
                if (e.data) {
                    data = e.data;
                }
                this.updateHelpBindingObject(data);
            });
        }
        this.hostComboLookupComponent.clear.subscribe(() => {
            this.updateHelpBindingObject(null);
        });
    }
    // 更新UISTATE
    updateHelpBindingObject(data) {
        let idField;
        let textField;
        let spliter = this.defaultSpliter;
        if (this.hostHelpComponent) {
            idField = this.hostHelpComponent.idField;
            textField = this.hostHelpComponent.textField;
            spliter = this.hostHelpComponent.multipleChoiceSeparator || this.defaultSpliter;
        }
        if (this.hostComboListComponent) {
            idField = this.hostComboListComponent.idField;
            textField = this.hostComboListComponent.textField;
        }
        if (this.hostComboLookupComponent) {
            idField = this.hostComboLookupComponent.idField;
            textField = this.hostComboLookupComponent.textField;
            spliter = this.hostComboLookupComponent.separator || this.defaultSpliter;
        }
        this.emitUiStateBinding(data, idField, textField, spliter);
    }
    emitUiStateBinding(data, idField, textField, spliter = ',') {
        const newObject = { key: null, value: null };
        if (data) {
            // const idField = this.hostHelpComponent.idField;
            // const textField = this.hostHelpComponent.textField;
            if (Array.isArray(data)) {
                newObject.key = data.map((d) => d[idField]).join(spliter);
                newObject.value = data.map((d) => {
                    if (textField.indexOf('.') > -1) {
                        return textField.split('.').reduce((r, c) => {
                            return r[c];
                        }, d);
                    }
                    else {
                        return d[textField];
                    }
                }).join(spliter);
            }
            else {
                if (idField) {
                    newObject.key = data[idField];
                }
                if (textField) {
                    if (textField.indexOf('.') > -1) {
                        newObject.value = textField.split('.').reduce((r, c) => {
                            return r[c];
                        }, data);
                    }
                    else {
                        newObject.value = data[textField];
                    }
                }
            }
        }
        this.UIStateBindingChange.emit(newObject);
    }
}
UIStateBindingDirective.decorators = [
    { type: Directive, args: [{
                selector: '[UIStateBinding]'
            },] }
];
/** @nocollapse */
UIStateBindingDirective.ctorParameters = () => [
    { type: ComboBoxComponent, decorators: [{ type: Host }, { type: Self }, { type: Optional }] },
    { type: LookupGridComponent, decorators: [{ type: Host }, { type: Self }, { type: Optional }] },
    { type: KeyValueDiffers, decorators: [{ type: Optional }] },
    { type: ComboListComponent, decorators: [{ type: Host }, { type: Self }, { type: Optional }] },
    { type: ComboLookupComponent, decorators: [{ type: Host }, { type: Self }, { type: Optional }] }
];
UIStateBindingDirective.propDecorators = {
    bindingObject: [{ type: Input, args: ['UIStateBinding',] }],
    UIStateBindingChange: [{ type: Output }],
    onValueChange: [{ type: HostListener, args: ['valueChange', ['$event'],] }]
};
