// @gsp-wf/rtdevkit/esm5/lib/runtime/wf-workitem-handle/wf-workitem-handle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter, ViewChild } from '@angular/core';
import { TaskService } from '../services/task.service';
import { CompleteContext } from '../../entity';
import { NotifyService } from '@farris/ui-notify';
import { TaskSignaddComponent } from '../components/task-signadd/task-signadd.component';
import { DialogComponent } from '@farris/ui-dialog';
import { from } from 'rxjs';
import { I18nService } from '../services/i18n.service';
var WfWorkitemHandleComponent = /** @class */ (function () {
    function WfWorkitemHandleComponent(taskSvc, notify, i18nService) {
        var _this = this;
        this.taskSvc = taskSvc;
        this.notify = notify;
        this.i18nService = i18nService;
        this.completeContext = new CompleteContext();
        // 加签选中人
        this.displayPlaceholder = "请选择"; //跳转的默认显示节点
        //跳转的默认显示节点
        this.hasDefaultBranch = true;
        /**
         * 任务办理动作列表
         */
        this.actions = new EventEmitter();
        /**
         * 任务办理完成事件
         */
        this.completed = new EventEmitter();
        // 当前选中动作和动作实现
        this.preFuncList = [];
        this.postFuncList = [];
        // 弹框高度设置
        this.height = 420;
        this.actionsTitle = {
            pass: '',
            back: '',
            refuse: '',
            more: '',
            viewLog: '',
            viewFlowChart: '',
            viewBizlog: ''
        };
        /** @type {?} */
        var titles = this.i18nService.getValue('static.taskActions');
        Object.keys(this.actionsTitle).forEach((/**
         * @param {?} key
         * @return {?}
         */
        function (key) {
            _this.actionsTitle[key] = titles[key];
        }));
    }
    Object.defineProperty(WfWorkitemHandleComponent.prototype, "workitemId", {
        get: /**
         * @return {?}
         */
        function () {
            return this.inWorkitemId;
        },
        /**
         * 任务Id
         */
        set: /**
         * 任务Id
         * @param {?} workitemId
         * @return {?}
         */
        function (workitemId) {
            this.inWorkitemId = (workitemId && workitemId.trim());
            if (this.inWorkitemId) {
                this.refreshActions();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * 刷新动作列表
     */
    /**
     * 刷新动作列表
     * @private
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.refreshActions = /**
     * 刷新动作列表
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.taskSvc.getWorkItemActions(this.inWorkitemId).subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            // 临时方案：oData不支持自定义function $expand展开，临时序列化处理
            _this.taskActions = re.map((/**
             * @param {?} action
             * @return {?}
             */
            function (action) {
                action.parameters = JSON.parse(action.parameters);
                return action;
            }));
            _this.taskActions = _this.taskActions.filter((/**
             * @param {?} a
             * @return {?}
             */
            function (a) {
                return a.code !== 'ViewBizLog'
                    && a.code !== 'ViewFlowChart'
                    && a.code !== 'ViewExecuteForm';
            }));
            _this.actions.emit(_this.taskActions);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.showError(error);
        }));
    };
    /**
     * 动作完成后事件
     * @param result 事件结果
     */
    /**
     * 动作完成后事件
     * @private
     * @param {?=} result 事件结果
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.completedEvt = /**
     * 动作完成后事件
     * @private
     * @param {?=} result 事件结果
     * @return {?}
     */
    function (result) {
        if (result === void 0) { result = true; }
        if (result) {
            this.refreshActions();
        }
        this.completed.emit(result);
    };
    /**
     * 完成审批任务
     * @param actionCode 任务办理动作编号
     * @param comment? 办理意见
     */
    /**
     * 完成审批任务
     * @param {?} actionCode 任务办理动作编号
     * @param {?=} comment
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.complete = /**
     * 完成审批任务
     * @param {?} actionCode 任务办理动作编号
     * @param {?=} comment
     * @return {?}
     */
    function (actionCode, comment) {
        var _this = this;
        switch (actionCode) {
            case 'AssignUser': { // 指派
                this.assignee(this.inWorkitemId);
                break;
            }
            case 'Claim': // 领用
                this.taskSvc.claim(this.inWorkitemId).subscribe((/**
                 * @param {?} r
                 * @return {?}
                 */
                function (r) {
                    _this.notify.success(_this.i18nService.getValue('cueWords.successClaim'));
                    _this.refreshActions();
                    _this.completed.emit(false);
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    _this.showError(error);
                }));
                break;
            case 'Unclaim': // 取消领用
                this.taskSvc.unClaim(this.inWorkitemId).subscribe((/**
                 * @param {?} r
                 * @return {?}
                 */
                function (r) {
                    _this.notify.success(_this.i18nService.getValue('cueWords.successUnClaim'));
                    _this.refreshActions();
                    _this.completed.emit(false);
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    _this.showError(error);
                }));
                break;
            case 'Retrieve': // 取回
                this.taskSvc.retrieve({ workItemId: this.inWorkitemId, comment: comment }).subscribe((/**
                 * @param {?} r
                 * @return {?}
                 */
                function (r) {
                    _this.notify.success(_this.i18nService.getValue('cueWords.successRetrieve'));
                    _this.completedEvt();
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    _this.showError(error);
                }));
                break;
            case 'Transfer': { // 移交
                this.transferContext = {
                    workItemId: this.inWorkitemId,
                    comment: comment
                };
                break;
            }
            case 'AddSignFront': // 征询
                this.addSignContext = {
                    workItemId: this.inWorkitemId,
                    action: 'AddSignFront'
                };
                break;
            default:
                this.executeTask(actionCode, comment);
        }
    };
    /**
     * @param {?} sucess
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.refreshList = /**
     * @param {?} sucess
     * @return {?}
     */
    function (sucess) {
        this.completedEvt(sucess);
    };
    // 弹出加签帮助
    // 弹出加签帮助
    /**
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.runAddSign = 
    // 弹出加签帮助
    /**
     * @return {?}
     */
    function () {
        this.addSignContext = {
            workItemId: this.inWorkitemId,
            action: 'AddSignBehind'
        };
    };
    // 加签完成
    // 加签完成
    /**
     * @param {?} sucess
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.addSignComplete = 
    // 加签完成
    /**
     * @param {?} sucess
     * @return {?}
     */
    function (sucess) {
        if (this.addSignContext.action === 'AddSignFront') {
            this.refreshList(sucess);
        }
        else {
            this.completeContext.signAssignees = this.addSignCmp.selectedUsers.map((/**
             * @param {?} user
             * @return {?}
             */
            function (user) { return user.ID; }));
            this.selectAddSignees = this.addSignCmp.selectedUsers.map((/**
             * @param {?} user
             * @return {?}
             */
            function (user) { return user.Name; })).join(',');
        }
    };
    // 审批确认
    // 审批确认
    /**
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.handleOk = 
    // 审批确认
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.completeContext.comment) {
            this.notify.info("请输入审批意见");
            return;
        }
        if (!this.hasDefaultBranch && this.completeContext.destinationActivityId.length <= 0) { //没有默认分支并且用户没选择跳转节点，不允许提交
            this.notify.info("请选择目标节点");
            return;
        }
        this.taskSvc.excuteTask(this.inWorkitemId, this.completeContext).subscribe((/**
         * @param {?} cr
         * @return {?}
         */
        function (cr) {
            _this.commentModal.close();
            // 完成工作项后，检查是否需要执行后事件
            _this.excutePostEvent();
            if (_this.completeContext.conclusionValue === 'Pass') {
                _this.notify.success(_this.i18nService.getValue('cueWords.successPass'));
            }
            else if (_this.completeContext.conclusionValue === 'Back') {
                _this.notify.success(_this.i18nService.getValue('cueWords.successBack'));
            }
            else if (_this.completeContext.conclusionValue === 'Refuse') {
                _this.notify.success(_this.i18nService.getValue('cueWords.successRefuse'));
            }
            else {
                _this.notify.success(_this.i18nService.getValue('cueWords.success'));
            }
            if (cr.needInteraction) {
                _this.refreshActions(); //通过后弹出指派 add by fpf
                _this.assigneeInfos = [cr.nextFlowNodes[0].assigneeInfo];
            }
            else {
                _this.completedEvt();
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.showError(error);
        }));
    };
    // 取消操作
    // 取消操作
    /**
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.handleCancel = 
    // 取消操作
    /**
     * @return {?}
     */
    function () {
        this.completeContext = new CompleteContext();
        this.commentModal.close();
    };
    // 指派操作
    // 指派操作
    /**
     * @private
     * @param {?} workItemId
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.assignee = 
    // 指派操作
    /**
     * @private
     * @param {?} workItemId
     * @return {?}
     */
    function (workItemId) {
        var _this = this;
        this.taskSvc.getAssigneeInfo(workItemId).subscribe((/**
         * @param {?} info
         * @return {?}
         */
        function (info) {
            info.assigneeInfo.workItemId = workItemId; //兼容处理，此处返回的workitemid为null
            _this.assigneeInfos = [info.assigneeInfo];
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.showError(error);
        }));
    };
    // 重新初始化当前参数
    // 重新初始化当前参数
    /**
     * @private
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.initState = 
    // 重新初始化当前参数
    /**
     * @private
     * @return {?}
     */
    function () {
        this.selectAddSignees = '';
        this.destActivities = null;
        this.needAddSign = false;
    };
    /**
     * @private
     * @param {?} actionCode
     * @param {?=} opinion
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.executeTask = /**
     * @private
     * @param {?} actionCode
     * @param {?=} opinion
     * @return {?}
     */
    function (actionCode, opinion) {
        var _this = this;
        this.initState();
        this.completeContext = {
            workItemId: this.inWorkitemId,
            conclusionValue: actionCode,
            comment: '',
            signAssignees: [],
            // 加签人员
            destinationActivityId: '' // 目标节点
        };
        this.completeContext.comment = actionCode === 'Pass' ? this.actionsTitle.pass : '';
        if (opinion) {
            this.completeContext.comment = opinion;
        }
        /** @type {?} */
        var action = this.taskActions.find((/**
         * @param {?} a
         * @return {?}
         */
        function (a) { return a.code === actionCode; }));
        this.taskSvc.getActionEntity(this.workitemId, actionCode).subscribe((/**
         * @param {?} act
         * @return {?}
         */
        function (act) {
            if (!act) { // 审批动作已删除，走完成工作项  TODO: 非完成工作项动作如何处理？
                _this.showExcuteModal(action);
            }
            else {
                from(_this.excuteAction(act)).subscribe((/**
                 * @return {?}
                 */
                function () {
                    if (!act.webComponent) {
                        _this.showExcuteModal(action);
                    }
                }));
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.showError(error);
        }));
    };
    /**
     * 加签，跳转情况
     * @param action 动作
     */
    /**
     * 加签，跳转情况
     * @private
     * @param {?} action 动作
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.showExcuteModal = /**
     * 加签，跳转情况
     * @private
     * @param {?} action 动作
     * @return {?}
     */
    function (action) {
        // 动作参数按照
        if (action && action.parameters) {
            // 跳转、驳回场景
            if (action.parameters.SelectTargetAble === 'True' || action.parameters.SelectTargetAble === 'true') {
                // this.taskSvc.getDestinationScope(this.inWorkitemId, action.code).subscribe((re: { id: string, name: string }[]) => {
                //   this.destActivities = re;
                // });
                this.getDestinationNodes(this.inWorkitemId, action.code);
            }
            // 需要加签
            if (action.parameters.AddSignBehindAble === 'True' || action.parameters.AddSignBehindAble === 'true') {
                this.needAddSign = true;
            }
        }
        this.commentModal.show();
    };
    /**
     * @private
     * @param {?} workItemId
     * @param {?} actionCode
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.getDestinationNodes = /**
     * @private
     * @param {?} workItemId
     * @param {?} actionCode
     * @return {?}
     */
    function (workItemId, actionCode) {
        var _this = this;
        /** @type {?} */
        var nextNodes = [];
        this.taskSvc.getDestinationScope(workItemId, actionCode).subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            nextNodes = re;
            /** @type {?} */
            var des = re.map((/**
             * @param {?} a
             * @return {?}
             */
            function (a) { return a.nodeInfoList; }));
            /** @type {?} */
            var list = [];
            des.forEach((/**
             * @param {?} i
             * @return {?}
             */
            function (i) {
                list = list.concat(i);
            }));
            _this.destActivities = list.map((/**
             * @param {?} i
             * @return {?}
             */
            function (i) { return ({ id: i.activityDefinitionId, name: i.activityDefinitionName }); }));
            _this.hasDefaultBranch = _this.hasDefaultBranchFunc(nextNodes);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.showError(error);
        }));
    };
    /**
     * 跳转，返回的数据是否有默认分支
     * 如果返回的没有默认节点，跳转节点又没选的话，不允许提交
     * @param nextNodes
     */
    /**
     * 跳转，返回的数据是否有默认分支
     * 如果返回的没有默认节点，跳转节点又没选的话，不允许提交
     * @private
     * @param {?} nextNodes
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.hasDefaultBranchFunc = /**
     * 跳转，返回的数据是否有默认分支
     * 如果返回的没有默认节点，跳转节点又没选的话，不允许提交
     * @private
     * @param {?} nextNodes
     * @return {?}
     */
    function (nextNodes) {
        for (var i = 0; i < nextNodes.length; i++) {
            if (nextNodes[i].defaultBranch) {
                this.displayPlaceholder = nextNodes[i].nodeInfoList[0].activityDefinitionName;
                return true;
            }
        }
        return false;
    };
    /**
     * @private
     * @param {?} act
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.excuteAction = /**
     * @private
     * @param {?} act
     * @return {?}
     */
    function (act) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _loop_1, this_1, i;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getActionWebEventsFunc(act)];
                    case 1:
                        _a.sent();
                        if (this.preFuncList && this.preFuncList.length > 0) {
                            _loop_1 = function (i) {
                                this_1.preFuncList.filter((/**
                                 * @param {?} item
                                 * @return {?}
                                 */
                                function (item) { return item.index === i; }))[0].func();
                            };
                            this_1 = this;
                            for (i = 0; i < this.preFuncList.length; i++) {
                                _loop_1(i);
                            }
                        }
                        if (this.webCptFunc) {
                            this.webCptFunc();
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @private
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.excutePostEvent = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.postFuncList && this.postFuncList.length > 0) {
            var _loop_2 = function (i) {
                this_2.postFuncList.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return item.index === i; }))[0].func();
            };
            var this_2 = this;
            for (var i = 0; i < this.postFuncList.length; i++) {
                _loop_2(i);
            }
        }
    };
    /**
     * 获取动作的前后事件和实现方法
     * @param act 动作
     */
    /**
     * 获取动作的前后事件和实现方法
     * @private
     * @param {?} act 动作
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.getActionWebEventsFunc = /**
     * 获取动作的前后事件和实现方法
     * @private
     * @param {?} act 动作
     * @return {?}
     */
    function (act) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var e_1, _a, e_2, _b, e_3, _c, reqMap, preIndex, postIndex, _d, _e, e, pre, cpt, _f, _g, a, post, reqMap_1, reqMap_1_1, r, m, e_3_1;
            return tslib_1.__generator(this, function (_h) {
                switch (_h.label) {
                    case 0:
                        this.preFuncList = [];
                        this.webCptFunc = null;
                        this.postFuncList = [];
                        reqMap = [];
                        preIndex = 0;
                        postIndex = 0;
                        if (act.preEvent && act.preEvent.length > 0) {
                            try {
                                for (_d = tslib_1.__values(act.preEvent), _e = _d.next(); !_e.done; _e = _d.next()) {
                                    e = _e.value;
                                    pre = {
                                        event: 'pre',
                                        index: preIndex++,
                                        funcName: e.component.split('#')[1],
                                        req: System.import(e.component.split('#')[0])
                                    };
                                    reqMap.push(pre);
                                }
                            }
                            catch (e_1_1) { e_1 = { error: e_1_1 }; }
                            finally {
                                try {
                                    if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
                                }
                                finally { if (e_1) throw e_1.error; }
                            }
                        }
                        if (act.webComponent) {
                            cpt = {
                                event: 'cpt',
                                index: 0,
                                funcName: act.webComponent.split('#')[1],
                                req: System.import(act.webComponent.split('#')[0])
                            };
                            reqMap.push(cpt);
                        }
                        if (act.postEvent && act.postEvent.length > 0) {
                            try {
                                for (_f = tslib_1.__values(act.postEvent), _g = _f.next(); !_g.done; _g = _f.next()) {
                                    a = _g.value;
                                    post = {
                                        event: 'post',
                                        index: postIndex++,
                                        funcName: a.component.split('#')[1],
                                        req: System.import(a.component.split('#')[0])
                                    };
                                    reqMap.push(post);
                                }
                            }
                            catch (e_2_1) { e_2 = { error: e_2_1 }; }
                            finally {
                                try {
                                    if (_g && !_g.done && (_b = _f.return)) _b.call(_f);
                                }
                                finally { if (e_2) throw e_2.error; }
                            }
                        }
                        _h.label = 1;
                    case 1:
                        _h.trys.push([1, 6, 7, 8]);
                        reqMap_1 = tslib_1.__values(reqMap), reqMap_1_1 = reqMap_1.next();
                        _h.label = 2;
                    case 2:
                        if (!!reqMap_1_1.done) return [3 /*break*/, 5];
                        r = reqMap_1_1.value;
                        return [4 /*yield*/, r.req];
                    case 3:
                        m = _h.sent();
                        if (r.event === 'pre') {
                            this.preFuncList.push({
                                index: r.index,
                                func: m.default[r.funcName]
                            });
                        }
                        else if (r.event === 'post') {
                            this.postFuncList.push({
                                index: r.index,
                                func: m.default[r.funcName]
                            });
                        }
                        else {
                            this.webCptFunc = m.default[r.funcName];
                        }
                        _h.label = 4;
                    case 4:
                        reqMap_1_1 = reqMap_1.next();
                        return [3 /*break*/, 2];
                    case 5: return [3 /*break*/, 8];
                    case 6:
                        e_3_1 = _h.sent();
                        e_3 = { error: e_3_1 };
                        return [3 /*break*/, 8];
                    case 7:
                        try {
                            if (reqMap_1_1 && !reqMap_1_1.done && (_c = reqMap_1.return)) _c.call(reqMap_1);
                        }
                        finally { if (e_3) throw e_3.error; }
                        return [7 /*endfinally*/];
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 显示报错信息
     */
    /**
     * 显示报错信息
     * @private
     * @param {?} error
     * @return {?}
     */
    WfWorkitemHandleComponent.prototype.showError = /**
     * 显示报错信息
     * @private
     * @param {?} error
     * @return {?}
     */
    function (error) {
        if (error && error.error && error.error.Message) {
            this.notify.error(error.error.Message);
        }
        else {
            this.notify.error(error);
        }
    };
    WfWorkitemHandleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-workitem-handle',
                    template: "<farris-dialog #commentModal [enableScroll]= \"false\" [buttons]=\"buttonRef\" [width]=\"550\" [height]=\"360\" [title]=\"'static.title' | translate\">\r\n    <div style=\"padding: 20px;\">\r\n        <div>\r\n            <wf-comment class=\"w-100\" [(ngModel)]=\"completeContext.comment\" name=\"comment\"\r\n                [placeholder]=\"'static.inputHolder' | translate\"></wf-comment>\r\n        </div>\r\n        <div class=\"d-flex\" style=\"height:28px;overflow: hidden;\">\r\n            <div *ngIf=\"needAddSign\" style=\"flex:1;display: flex;height:28px;overflow: hidden;\" [ngStyle]=\"{'padding-right': destActivities ? '20px' : '0px'}\">\r\n                <label style=\"padding-right: 10px;line-height: 24px;\">{{ 'static.addSignee' | translate }}</label>\r\n                <div class=\"d-flex\" style=\"flex: 1\">\r\n                    <div class=\"input-group\">\r\n                        <input type=\"text\" class=\"form-control\" [(ngModel)]=\"selectAddSignees\" [readOnly]=\"true\" placeholder=\"\u8BF7\u9009\u62E9\"/>\r\n                        <button class=\"k-button k-bare\" type=\"button\" (click)=\"runAddSign()\">\r\n                            <i class=\"f-icon f-icon-lookup\" style=\"color: #bfbfbf; margin: 0 -8px 0 0;\"></i>\r\n                        </button>\r\n                    </div>\r\n                </div>\r\n            </div>\r\n            <div *ngIf=\"destActivities\" style=\"flex:1;display: flex;height:28px;overflow: hidden;\">\r\n                <label style=\"padding-right: 10px;line-height: 24px;\">{{ 'static.selectActivity' | translate }}</label>\r\n                <div class=\"d-flex\" style=\"flex:1\">\r\n                    <farris-combo-list  [data]=\"destActivities\" [textField]=\"'name'\" [valueField]=\"'id'\" [idField]=\"'id'\"\r\n                        [(ngModel)]=\"completeContext.destinationActivityId\" style=\"flex: 1;\" [placeholder] = \"displayPlaceholder\">\r\n                    </farris-combo-list >\r\n                </div>\r\n            </div>\r\n        </div>\r\n    </div>\r\n</farris-dialog>\r\n<ng-template #buttonRef>\r\n    <button class=\"btn btn-sm btn-primary px-3\" (click)=\"handleOk()\">{{\r\n            'static.footOk' | translate }}</button>\r\n    <button class=\"btn btn-sm btn-default px-3\" (click)=\"handleCancel()\">{{\r\n            'static.footCancel' | translate }}</button>\r\n</ng-template>\r\n\r\n<!-- \u6307\u6D3E -->\r\n<wf-task-assign [assigneeInfo]=\"assigneeInfos\" (completed)=\"refreshList($event)\"></wf-task-assign>\r\n<!-- \u79FB\u4EA4 -->\r\n<wf-task-transfer [transfer_context]=\"transferContext\" (completed)=\"refreshList($event)\"></wf-task-transfer>\r\n<!-- \u52A0\u7B7E\u3001\u5F81\u8BE2 -->\r\n<wf-task-signadd [addsign_context]=\"addSignContext\" (completed)=\"addSignComplete($event)\"></wf-task-signadd>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    WfWorkitemHandleComponent.ctorParameters = function () { return [
        { type: TaskService },
        { type: NotifyService },
        { type: I18nService }
    ]; };
    WfWorkitemHandleComponent.propDecorators = {
        addSignCmp: [{ type: ViewChild, args: [TaskSignaddComponent,] }],
        commentModal: [{ type: ViewChild, args: ['commentModal',] }],
        workitemId: [{ type: Input }],
        actions: [{ type: Output }],
        completed: [{ type: Output }]
    };
    return WfWorkitemHandleComponent;
}());
export { WfWorkitemHandleComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    WfWorkitemHandleComponent.prototype.inWorkitemId;
    /**
     * @type {?}
     * @private
     */
    WfWorkitemHandleComponent.prototype.taskActions;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.completeContext;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.assigneeInfos;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.transferContext;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.destActivities;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.addSignContext;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.needAddSign;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.selectAddSignees;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.displayPlaceholder;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.hasDefaultBranch;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.addSignCmp;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.commentModal;
    /**
     * 任务办理动作列表
     * @type {?}
     */
    WfWorkitemHandleComponent.prototype.actions;
    /**
     * 任务办理完成事件
     * @type {?}
     */
    WfWorkitemHandleComponent.prototype.completed;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.preFuncList;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.webCptFunc;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.postFuncList;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.height;
    /** @type {?} */
    WfWorkitemHandleComponent.prototype.actionsTitle;
    /**
     * @type {?}
     * @private
     */
    WfWorkitemHandleComponent.prototype.taskSvc;
    /**
     * @type {?}
     * @private
     */
    WfWorkitemHandleComponent.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    WfWorkitemHandleComponent.prototype.i18nService;
}
