// @progress/kendo-angular-dialog/dist/es/window/actions/window-maximize-action.directive.js
import * as tslib_1 from "tslib";
import { Directive, HostListener, ElementRef, Renderer2, Optional, HostBinding, Input } from "@angular/core";
import { DragResizeService } from './../drag-resize.service';
import { Button } from '@progress/kendo-angular-buttons';
import { L10N_PREFIX, LocalizationService } from '@progress/kendo-angular-l10n';
var WindowMaximizeActionDirective = /** @class */ (function (_super) {
    tslib_1.__extends(WindowMaximizeActionDirective, _super);
    function WindowMaximizeActionDirective(el, renderer, _service, localization) {
        var _this = _super.call(this, el, renderer, null, localization) || this;
        _this.window = _service;
        _this.look = 'bare';
        _this.icon = 'window-maximize';
        return _this;
    }
    WindowMaximizeActionDirective.prototype.onClick = function () {
        if (!this.isDisabled) {
            this.window.maximizeAction();
        }
    };
    Object.defineProperty(WindowMaximizeActionDirective.prototype, "visible", {
        get: function () {
            return this.window.options.state === 'default' ? 'inline-flex' : 'none';
        },
        enumerable: true,
        configurable: true
    });
    WindowMaximizeActionDirective.decorators = [
        { type: Directive, args: [{
                    exportAs: 'kendoWindowMaximizeAction',
                    providers: [
                        LocalizationService,
                        {
                            provide: L10N_PREFIX,
                            useValue: 'kendo.button'
                        }
                    ],
                    selector: 'button[kendoWindowMaximizeAction]' // tslint:disable-line
                },] },
    ];
    /** @nocollapse */
    WindowMaximizeActionDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
        { type: DragResizeService, decorators: [{ type: Optional },] },
        { type: LocalizationService, },
    ]; };
    WindowMaximizeActionDirective.propDecorators = {
        'window': [{ type: Input },],
        'onClick': [{ type: HostListener, args: ['click',] },],
        'visible': [{ type: HostBinding, args: ['style.display',] },],
    };
    return WindowMaximizeActionDirective;
}(Button));
export { WindowMaximizeActionDirective };
