// @progress/kendo-angular-dialog/dist/es/window/actions/window-restore-action.directive.js
import * as tslib_1 from "tslib";
import { Directive, HostListener, ElementRef, Renderer2, Optional, HostBinding, Input } from "@angular/core";
import { DragResizeService } from './../drag-resize.service';
import { Button } from '@progress/kendo-angular-buttons';
import { L10N_PREFIX, LocalizationService } from '@progress/kendo-angular-l10n';
var WindowRestoreActionDirective = /** @class */ (function (_super) {
    tslib_1.__extends(WindowRestoreActionDirective, _super);
    function WindowRestoreActionDirective(el, renderer, _service, localization) {
        var _this = _super.call(this, el, renderer, null, localization) || this;
        _this.window = _service;
        _this.look = 'bare';
        _this.icon = 'window-restore';
        return _this;
    }
    WindowRestoreActionDirective.prototype.onClick = function () {
        if (!this.isDisabled) {
            this.window.restoreAction();
        }
    };
    Object.defineProperty(WindowRestoreActionDirective.prototype, "visible", {
        get: function () {
            return this.window.options.state === 'default' ? 'none' : 'inline-flex';
        },
        enumerable: true,
        configurable: true
    });
    WindowRestoreActionDirective.decorators = [
        { type: Directive, args: [{
                    exportAs: 'kendoWindowRestoreAction',
                    providers: [
                        LocalizationService,
                        {
                            provide: L10N_PREFIX,
                            useValue: 'kendo.button'
                        }
                    ],
                    selector: 'button[kendoWindowRestoreAction]' // tslint:disable-line
                },] },
    ];
    /** @nocollapse */
    WindowRestoreActionDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
        { type: DragResizeService, decorators: [{ type: Optional },] },
        { type: LocalizationService, },
    ]; };
    WindowRestoreActionDirective.propDecorators = {
        'window': [{ type: Input },],
        'onClick': [{ type: HostListener, args: ['click',] },],
        'visible': [{ type: HostBinding, args: ['style.display',] },],
    };
    return WindowRestoreActionDirective;
}(Button));
export { WindowRestoreActionDirective };
