// @gsp-sys/sysmgr-ui/esm2015/lib/sys-manager-ui/group-nav/group.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { SysCommonUtilService, SysDataUtil } from '@gsp-sys/sysmgr-common';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "@gsp-sys/sysmgr-common";
export class GroupService {
    /**
     * @param {?} http
     * @param {?} sysUtil
     */
    constructor(http, sysUtil) {
        this.http = http;
        this.sysUtil = sysUtil;
    }
    /**
     * @private
     * @param {?} type
     * @return {?}
     */
    getUrlByGroupType(type) {
        /** @type {?} */
        let url = '';
        switch (type) {
            case 'user':
                url = this.sysUtil.userGroupWebApi;
                break;
            case 'position':
                url = this.sysUtil.posGroupWebApi;
                break;
            case 'role':
                url = this.sysUtil.roleGroupWebApi;
                break;
            case 'org':
                url = this.sysUtil.sysOrgWebApi;
                break;
            // 默认用户分组
            default:
                url = this.sysUtil.userGroupWebApi;
                break;
        }
        return url;
    }
    /**
     * 根据级数获取组织
     * @param {?} layer 级数
     * @param {?} type
     * @param {?} filter
     * @return {?}
     */
    getByLayer(layer, type, filter) {
        /** @type {?} */
        const url = this.getUrlByGroupType(type);
        if (!filter) {
            filter = {};
        }
        filter.layer = layer;
        filter.parentId = null;
        SysDataUtil.deleteEmpty(filter);
        /** @type {?} */
        const opt = {
            headers: this.sysUtil.getHeader().headers,
            params: { param: JSON.stringify(filter) }
        };
        return this.http.get(url, opt);
    }
    /**
     * 根据级数获取组织
     * @param {?} parentId 父节点
     * @param {?} type
     * @param {?} filter
     * @return {?}
     */
    getByParentId(parentId, type, filter) {
        /** @type {?} */
        const url = this.getUrlByGroupType(type);
        if (!filter) {
            filter = {};
        }
        filter.layer = null;
        filter.parentId = parentId;
        SysDataUtil.deleteEmpty(filter);
        /** @type {?} */
        const opt = {
            headers: this.sysUtil.getHeader().headers,
            params: { param: JSON.stringify(filter) }
        };
        return this.http.get(url, opt);
    }
    /**
     * 根据级数获取组织
     * @param {?} type
     * @param {?} filter
     * @return {?}
     */
    searchData(type, filter) {
        /** @type {?} */
        const url = this.getUrlByGroupType(type);
        if (!filter) {
            filter = {};
        }
        filter.layer = null;
        filter.parentId = null;
        SysDataUtil.deleteEmpty(filter);
        /** @type {?} */
        const opt = {
            headers: this.sysUtil.getHeader().headers,
            params: { param: JSON.stringify(filter) }
        };
        return this.http.get(url, opt);
    }
}
GroupService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
GroupService.ctorParameters = () => [
    { type: HttpClient },
    { type: SysCommonUtilService }
];
/** @nocollapse */ GroupService.ngInjectableDef = i0.defineInjectable({ factory: function GroupService_Factory() { return new GroupService(i0.inject(i1.HttpClient), i0.inject(i2.SysCommonUtilService)); }, token: GroupService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    GroupService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    GroupService.prototype.sysUtil;
}
