// @gsp-sys/sysmgr-ui/esm2015/lib/sys-manager-ui/group-nav/group-nav.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Input, EventEmitter, Output, HostListener, ElementRef } from '@angular/core';
import { TreeTableComponent } from '@farris/ui-treetable';
import { from } from 'rxjs';
import { GroupService } from './group.service';
import { filter } from 'rxjs/operators';
import { SysLocaleService } from '../../locale/sys-locale.service';
import { NAVI_LOCALE_ZHCHS } from '../../locale/locales/zh-cn/navigiation';
import { MessagerService } from '@farris/ui-messager';
import { NotifyService } from '@farris/ui-notify';
export class GroupNavComponent {
    /**
     * @param {?} groupService
     * @param {?} localService
     * @param {?} notifyService
     * @param {?} msg
     */
    constructor(groupService, localService, notifyService, msg) {
        this.groupService = groupService;
        this.localService = localService;
        this.notifyService = notifyService;
        this.msg = msg;
        this.showHeader = true;
        this.fixedHeader = true;
        this.withCodeColumn = true;
        // 是否添加所有根节点
        this.allRoot = false;
        this.selectKeys = [];
        this.needRemote = true;
        /**
         * 显示组织类型列
         */
        this.showNodeTypeColumn = false;
        this.showFilterBar = false;
        this.searchFields = [{ label: "编号", value: "code" }, { label: "名称", value: "name" }];
        this.allColumnsTitle = '所有列';
        this.search = new EventEmitter();
        this.selectRowChanged = new EventEmitter();
        /**
         * 查询内容
         */
        this.searchInfo = '';
        /**
         * 上次查询内容
         */
        this.oldSearchInfo = '';
        this.field = '';
        /**
         * 上次查询内容
         */
        this.oldField = '';
        //上次搜索结果
        this.searchResult = [];
        //上次搜索的位置
        this.searchIndex = 0;
        /**
         * 监控拖拽事件，重绘组织TreeTable
         * @param event 事件参数
         */
        this.temp = true;
    }
    /**
     * @return {?}
     */
    winResize() {
        this.reSizeTreeTable();
    }
    /**
     * 重绘Farris-TreeTable的大小
     * @param {?=} height
     * @param {?=} width
     * @return {?}
     */
    reSizeTreeTable(height = 0, width = 0) {
        if (height === 0) {
            height = this.farrisTreeParent.nativeElement.clientHeight;
        }
        if (width === 0) {
            width = this.farrisTreeParent.nativeElement.clientWidth;
        }
        this.farrisTt.resize({ height, width });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.localService) {
            this.i18nRes = this.localService.getComponentOpt('navi');
        }
        else {
            this.i18nRes = NAVI_LOCALE_ZHCHS;
        }
        this.searchFields = [{ label: this.i18nRes.codeField, value: "code" }, { label: this.i18nRes.nameField, value: "name" }];
        this.allColumnsTitle = this.i18nRes.allColumnsTitle;
        // 启用编号列
        if (this.withCodeColumn) {
            this.fixedHeader = true;
            this.showHeader = true;
            this.cols = [
                { field: 'code', title: this.i18nRes.codeField, width: 100 },
                { field: 'name', title: this.i18nRes.nameField, width: 100 }
            ];
        }
        else {
            this.fixedHeader = false;
            this.showHeader = false;
            this.cols = [
                { field: 'name', title: this.i18nRes.nameField, width: 100 }
            ];
        }
        // 需要获取数据，否则直接根据treeData初始数据
        if (this.needRemote) {
            // 所有标识
            if (this.allRoot) {
                this.treeData = [];
                /** @type {?} */
                const rootNode = {};
                rootNode.children = [];
                rootNode.data = {
                    id: 'all',
                    code: this.i18nRes.rootNodeCode,
                    name: this.i18nRes.rootNodeTitle
                };
                rootNode.expanded = true;
                rootNode.leaf = false;
                rootNode.selectable = true;
                this.treeData.push(rootNode);
            }
            else {
                this.treeData = [];
            }
            this.groupService.getByLayer(1, this.groupType, this.filter).subscribe((/**
             * @param {?} val
             * @return {?}
             */
            val => {
                this.initTreeNodeList(val);
            }));
        }
        else { // 不需要从服务器端获取
            this.farrisTt.updateSerializedValue();
        }
        // 异步加载，通过children为0来判断从未展开过节点
        this.farrisTt.expand.pipe(filter((/**
         * @param {?} node
         * @return {?}
         */
        (node) => node.children.length === 0))).subscribe((/**
         * @param {?} node
         * @return {?}
         */
        (node) => {
            node['showLoading'] = true;
            /** @type {?} */
            const id = node.data.id;
            // 获取系统组织
            this.groupService.getByParentId(id, this.groupType, this.filter).subscribe((/**
             * @param {?} list
             * @return {?}
             */
            (list) => {
                node['showLoading'] = false;
                /** @type {?} */
                const childNode = [];
                // 转换下级组织
                from(list).subscribe((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => {
                    childNode.push(this.convertToTreeNode(item));
                }));
                // 防止出现下级节点没有数据
                if (childNode.length > 0) {
                    this.farrisTt.appendChildren(childNode, node);
                }
                else { // 应该是无下级权限才走这里
                    node.leaf = true;
                    this.farrisTt.updateSerializedValue();
                }
            }));
        }));
    }
    /**
     * 刷新当前页面信息
     * @return {?}
     */
    refreshThisNav() {
        this.treeData = [];
        /** @type {?} */
        const rootNode = {};
        rootNode.children = [];
        rootNode.data = {
            id: 'all',
            code: this.i18nRes.rootNodeCode,
            name: this.i18nRes.rootNodeTitle
        };
        rootNode.expanded = true;
        rootNode.leaf = false;
        rootNode.selectable = true;
        this.treeData.push(rootNode);
        this.groupService.getByLayer(1, this.groupType, this.filter).subscribe((/**
         * @param {?} val
         * @return {?}
         */
        val => {
            this.initTreeNodeList(val);
        }));
    }
    /**
     * 页面渲染完成后
     * @return {?}
     */
    ngAfterContentInit() {
        this.reSizeTreeTable();
    }
    /**
     * 搜索， 此处是分层加载，搜索时需要根据原数据进行搜索，搜索后，需要将对应父节点展示
     * @param {?} args  格式 searchData = { field: '*', value: '' };
     * @return {?}
     */
    searchData(args) {
        this.search.next(args);
        if (!this.filter) {
            this.filter = {};
        }
        if (!args) {
            return;
        }
        this.searchInfo = args.value;
        this.field = args.field;
        // 搜索内容为空
        if (!this.searchInfo || this.searchInfo.length <= 0) {
            this.clearSearchValue();
            return;
        }
        // 搜索内容与上次不同，重新搜索
        if (this.oldSearchInfo !== this.searchInfo || this.oldField !== this.field) {
            this.filter.isSearchTree = true;
            if (this.field === '*') {
                this.filter.codeOrName = this.searchInfo;
                this.filter.code = '';
                this.filter.name = '';
            }
            else if (this.field === 'code') {
                this.filter.code = this.searchInfo;
                this.filter.name = '';
                this.filter.codeOrName = '';
            }
            else if (this.field === 'name') {
                this.filter.name = this.searchInfo;
                this.filter.code = '';
                this.filter.codeOrName = '';
            }
            this.groupService.searchData(this.groupType, this.filter).subscribe((/**
             * @param {?} list
             * @return {?}
             */
            list => {
                this.treeData = [];
                this.searchResult = [];
                this.searchIndex = 0;
                if (list && list.length > 0) {
                    this.treeData = this.makeTree(list);
                    list.forEach((/**
                     * @param {?} item
                     * @return {?}
                     */
                    item => {
                        if (item.search_data) {
                            this.searchResult.push(item.id);
                        }
                    }));
                    // 手动更新TreeTable的值
                    this.farrisTt.updateSerializedValue();
                    setTimeout((/**
                     * @return {?}
                     */
                    () => {
                        this.farrisTt.selectNode(this.searchResult[0]);
                        this.farrisTt.collapseAll();
                        this.expandParent(this.searchResult[0]);
                        this.farrisTt.scrollToCurrentNode();
                    }));
                }
                else {
                    // 手动更新TreeTable的值
                    this.farrisTt.updateSerializedValue();
                    this.notifyService.info(this.i18nRes.searchNullData);
                }
            }));
        }
        else {
            if (this.searchResult && this.searchResult.length > 0) {
                this.searchIndex = this.searchIndex + 1;
                // 已经搜索到最后，需要重头搜索
                if (this.searchIndex === this.searchResult.length) {
                    this.searchIndex = 0;
                }
                setTimeout((/**
                 * @return {?}
                 */
                () => {
                    this.farrisTt.selectNode(this.searchResult[this.searchIndex]);
                    this.farrisTt.collapseAll();
                    this.expandParent(this.searchResult[this.searchIndex]);
                    this.farrisTt.scrollToCurrentNode();
                }));
            }
            else {
                this.notifyService.info(this.i18nRes.searchNullData);
            }
        }
        this.oldSearchInfo = this.searchInfo;
        this.oldField = this.field;
    }
    /**
     * @return {?}
     */
    clearSearchValue() {
        //点击清除时，防止事件执行两次
        if (!this.temp) {
            return;
        }
        this.temp = false;
        // 所有标识
        if (this.allRoot) {
            this.treeData = [];
            /** @type {?} */
            const rootNode = {};
            rootNode.children = [];
            rootNode.data = {
                id: 'all',
                code: this.i18nRes.rootNodeCode,
                name: this.i18nRes.rootNodeTitle
            };
            rootNode.expanded = true;
            rootNode.leaf = false;
            rootNode.selectable = true;
            this.treeData.push(rootNode);
        }
        else {
            this.treeData = [];
        }
        this.filter.isSearchTree = false;
        this.filter.codeOrName = null;
        this.filter.code = null;
        this.filter.name = null;
        this.groupService.getByLayer(1, this.groupType, this.filter).subscribe((/**
         * @param {?} val
         * @return {?}
         */
        val => {
            this.initTreeNodeList(val);
            if (this.treeData.length > 0) {
                this.farrisTt.selectNode(this.treeData[0].data.id);
            }
            this.temp = true;
        }));
        this.oldSearchInfo = '';
        this.oldField = '';
    }
    /**
     * 构造树
     * @private
     * @param {?} data
     * @return {?}
     */
    makeTree(data) {
        /** @type {?} */
        const root = data.filter((/**
         * @param {?} t
         * @return {?}
         */
        t => t.layer === 1)).map((/**
         * @param {?} t
         * @return {?}
         */
        t => {
            /** @type {?} */
            const node = {
                data: t,
                children: [],
                expanded: false,
                selectable: t.farris_selectable
            };
            if (!t.isDetail) {
                node.leaf = false;
                node.children = [];
            }
            return node;
        }));
        root.forEach((/**
         * @param {?} e
         * @return {?}
         */
        e => {
            // 构造下级：children，父级的id信息
            e.children = this.makeChildrenByParentId(data, e.data.id);
        }));
        return root;
    }
    /**
     * 父节点构造树
     * @private
     * @param {?} datas
     * @param {?} parentId
     * @return {?}
     */
    makeChildrenByParentId(datas, parentId) {
        /** @type {?} */
        const x = datas.filter((/**
         * @param {?} c
         * @return {?}
         */
        c => c.parentId === parentId)).map((/**
         * @param {?} d
         * @return {?}
         */
        d => {
            return {
                data: d,
                children: this.makeChildrenByParentId(datas, d.id),
                expanded: false,
                selectable: d.farris_selectable
            };
        }));
        return x;
    }
    /**
     * 将组织实体转换为Farris-TreeNode
     * @private
     * @param {?} src
     * @return {?}
     */
    convertToTreeNode(src) {
        /** @type {?} */
        const node = { data: src };
        // 非明细节点
        if (!src.isDetail) {
            node.leaf = false;
            node.children = [];
        }
        // TODO: 只有有权限的组织才能被选中,还是这里不做处理，发送事件时去处理
        node.selectable = src.farris_selectable;
        return node;
    }
    /**
     * 初始顶级分组，默认layer=1的情况
     * @private
     * @param {?} list
     * @return {?}
     */
    initTreeNodeList(list) {
        if (!list || list.length === 0) {
            return;
        }
        from(list).subscribe((/**
         * @param {?} firstLayer
         * @return {?}
         */
        firstLayer => {
            /** @type {?} */
            const node = this.convertToTreeNode(firstLayer);
            // 非明细，带往下多构造一级，假设有数据
            if (node.leaf) {
                /** @type {?} */
                const curId = firstLayer.id;
                /** @type {?} */
                const childern = list.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => item.parentId === curId));
                from(childern).subscribe((/**
                 * @param {?} secondLayer
                 * @return {?}
                 */
                secondLayer => {
                    node.children.push(this.convertToTreeNode(secondLayer));
                }));
            }
            // 如果添加所有根节点
            if (this.allRoot) {
                this.treeData[0].children.push(node);
            }
            else {
                this.treeData.push(node);
            }
        }));
        // 手动更新TreeTable的值
        this.farrisTt.updateSerializedValue();
    }
    /**
     * 选择的组织行
     * @param {?} event
     * @return {?}
     */
    onNodeSelect(event) {
        console.log(event.node.data);
        this.selectRowChanged.next(event.node.data);
    }
    /**
     * 获取当前TreeTable的绑定数据
     * @return {?}
     */
    getTreeData() {
        return this.treeData;
    }
    /**
     * 获取当前TreeTable的组件实例
     * @return {?}
     */
    getTreeTableComp() {
        return this.farrisTt;
    }
    /**
     * 选中某个节点
     * @param {?} id
     * @param {?=} isCheck
     * @return {?}
     */
    selectNode(id, isCheck) {
        this.farrisTt.selectNode(id, isCheck);
    }
    /**
     * 清除所有选中值
     * @return {?}
     */
    unCheckAllNodes() {
        this.farrisTt.unCheckAllNodes();
    }
    /**
     * 清除所有选中值
     * @param {?} key
     * @return {?}
     */
    scrollToCurrentNode(key) {
        if (key) {
            this.farrisTt.selectNode(key);
            this.expandParent(key);
            this.farrisTt.scrollToCurrentNode();
        }
    }
    /**
     * 逐一找到需要展开的父节点，全部展开
     * @param {?} key
     * @return {?}
     */
    expandParent(key) {
        /** @type {?} */
        const parent = this.farrisTt.findRowNode(key);
        /** @type {?} */
        let parentNode = parent.parent;
        if (parentNode) {
            while (parentNode) {
                this.farrisTt.expandNode(parentNode.id);
                parentNode = parentNode.parent;
            }
        }
        else {
            // this.farrisTt.expandNode(key);
        }
    }
}
GroupNavComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-group-nav',
                template: "<div #farrisTreeParent class=\"f-utils-fill-flex-column\" style=\"background-color: white\">\r\n    <farris-treetable #farrisTt \r\n    [data]=\"this.treeData\" \r\n    [columns]=\"cols\" \r\n    [idField]=\"'id'\" \r\n    [fixedHeader]=\"fixedHeader\" \r\n    [showHeader]=\"showHeader\" \r\n    [fit]=\"true\" \r\n    [resizableColumns]=\"true\"\r\n    [fitColumns]=\"true\"\r\n    [(showFilterBar)]=\"showFilterBar\"\r\n    (search)=\"searchData($event)\"\r\n    [searchFields]=\"searchFields\"\r\n    [allColumnsTitle]=\"allColumnsTitle\"\r\n    (clearSearchValue)=\"clearSearchValue()\"\r\n    (nodeSelected)=\"onNodeSelect($event)\">\r\n    </farris-treetable>\r\n</div>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
GroupNavComponent.ctorParameters = () => [
    { type: GroupService },
    { type: SysLocaleService },
    { type: NotifyService },
    { type: MessagerService }
];
GroupNavComponent.propDecorators = {
    farrisTreeParent: [{ type: ViewChild, args: ['farrisTreeParent',] }],
    farrisTt: [{ type: ViewChild, args: ['farrisTt',] }],
    title: [{ type: Input }],
    showHeader: [{ type: Input }],
    fixedHeader: [{ type: Input }],
    withCodeColumn: [{ type: Input }],
    allRoot: [{ type: Input }],
    groupType: [{ type: Input }],
    filter: [{ type: Input }],
    selectKeys: [{ type: Input }],
    needRemote: [{ type: Input }],
    showNodeTypeColumn: [{ type: Input }],
    treeData: [{ type: Input }],
    showFilterBar: [{ type: Input }],
    search: [{ type: Output }],
    selectRowChanged: [{ type: Output }],
    winResize: [{ type: HostListener, args: ['window:resize', [],] }]
};
if (false) {
    /** @type {?} */
    GroupNavComponent.prototype.farrisTreeParent;
    /** @type {?} */
    GroupNavComponent.prototype.farrisTt;
    /** @type {?} */
    GroupNavComponent.prototype.title;
    /** @type {?} */
    GroupNavComponent.prototype.showHeader;
    /** @type {?} */
    GroupNavComponent.prototype.fixedHeader;
    /** @type {?} */
    GroupNavComponent.prototype.withCodeColumn;
    /** @type {?} */
    GroupNavComponent.prototype.allRoot;
    /** @type {?} */
    GroupNavComponent.prototype.groupType;
    /** @type {?} */
    GroupNavComponent.prototype.filter;
    /** @type {?} */
    GroupNavComponent.prototype.selectKeys;
    /** @type {?} */
    GroupNavComponent.prototype.needRemote;
    /**
     * 显示组织类型列
     * @type {?}
     */
    GroupNavComponent.prototype.showNodeTypeColumn;
    /** @type {?} */
    GroupNavComponent.prototype.treeData;
    /** @type {?} */
    GroupNavComponent.prototype.showFilterBar;
    /** @type {?} */
    GroupNavComponent.prototype.searchFields;
    /** @type {?} */
    GroupNavComponent.prototype.allColumnsTitle;
    /** @type {?} */
    GroupNavComponent.prototype.search;
    /** @type {?} */
    GroupNavComponent.prototype.selectRowChanged;
    /** @type {?} */
    GroupNavComponent.prototype.cols;
    /** @type {?} */
    GroupNavComponent.prototype.i18nRes;
    /**
     * 查询内容
     * @type {?}
     */
    GroupNavComponent.prototype.searchInfo;
    /**
     * 上次查询内容
     * @type {?}
     */
    GroupNavComponent.prototype.oldSearchInfo;
    /** @type {?} */
    GroupNavComponent.prototype.field;
    /**
     * 上次查询内容
     * @type {?}
     */
    GroupNavComponent.prototype.oldField;
    /** @type {?} */
    GroupNavComponent.prototype.searchResult;
    /** @type {?} */
    GroupNavComponent.prototype.searchIndex;
    /**
     * 监控拖拽事件，重绘组织TreeTable
     * \@param event 事件参数
     * @type {?}
     */
    GroupNavComponent.prototype.temp;
    /** @type {?} */
    GroupNavComponent.prototype.groupService;
    /** @type {?} */
    GroupNavComponent.prototype.localService;
    /**
     * @type {?}
     * @private
     */
    GroupNavComponent.prototype.notifyService;
    /** @type {?} */
    GroupNavComponent.prototype.msg;
}
