// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/param-data-serializer/default-data-row-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/param-data-serializer/default-data-row-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DefaultDataRow } from '../../param-data-vo/default-data-row';
import { DataItem } from '../../param-data-vo/data-item';
var DefaultDataRowSerializer = /** @class */ (function () {
    function DefaultDataRowSerializer() {
    }
    /**
     * @param {?} dataRow
     * @return {?}
     */
    DefaultDataRowSerializer.prototype.serialize = /**
     * @param {?} dataRow
     * @return {?}
     */
    function (dataRow) {
        /** @type {?} */
        var jobj = new Object();
        if (!dataRow) {
            return null;
        }
        /** @type {?} */
        var defaultDataRow = (/** @type {?} */ (dataRow));
        jobj['id'] = defaultDataRow.id;
        jobj['kind'] = defaultDataRow.kind;
        if (defaultDataRow.paramItems) {
            jobj['paramItems'] = [];
            defaultDataRow.paramItems.forEach((/**
             * @param {?} paramItem
             * @return {?}
             */
            function (paramItem) {
                jobj['paramItems'] = paramItem;
            }));
        }
        return jobj;
    };
    /**
     * @param {?} jobj
     * @return {?}
     */
    DefaultDataRowSerializer.prototype.deserialize = /**
     * @param {?} jobj
     * @return {?}
     */
    function (jobj) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        var dataRow = new DefaultDataRow();
        dataRow.id = jobj['id'];
        if (jobj['paramItems']) {
            dataRow.paramItems = new Array();
            jobj['paramItems'].forEach((/**
             * @param {?} paramItemJobj
             * @return {?}
             */
            function (paramItemJobj) {
                dataRow.paramItems.push(Object.assign(new DataItem(), paramItemJobj));
            }));
        }
        return dataRow;
    };
    return DefaultDataRowSerializer;
}());
export { DefaultDataRowSerializer };
