// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/param-data-serializer/dimension-data-row-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/param-data-serializer/dimension-data-row-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DimensionDataRow } from '../../param-data-vo/dimension-data-row';
import { DataItem } from '../../param-data-vo/data-item';
var DimensionDataRowSerializer = /** @class */ (function () {
    function DimensionDataRowSerializer() {
    }
    /**
     * @param {?} dataRow
     * @return {?}
     */
    DimensionDataRowSerializer.prototype.serialize = /**
     * @param {?} dataRow
     * @return {?}
     */
    function (dataRow) {
        /** @type {?} */
        var jobj = new Object();
        if (!dataRow) {
            return null;
        }
        /** @type {?} */
        var dimensionDataRow = (/** @type {?} */ (dataRow));
        jobj['id'] = dimensionDataRow.id;
        jobj['kind'] = dimensionDataRow.kind;
        if (dimensionDataRow.dimensionDataItems) {
            jobj['dimensionDataItems'] = [];
            dimensionDataRow.dimensionDataItems.forEach((/**
             * @param {?} dimensionDataItem
             * @return {?}
             */
            function (dimensionDataItem) {
                jobj['dimensionDataItems'] = dimensionDataItem;
            }));
        }
        if (dimensionDataRow.paramDataItems) {
            jobj['paramDataItems'] = [];
            dimensionDataRow.paramDataItems.forEach((/**
             * @param {?} paramItem
             * @return {?}
             */
            function (paramItem) {
                jobj['paramDataItems'] = paramItem;
            }));
        }
        return jobj;
    };
    /**
     * @param {?} jobj
     * @return {?}
     */
    DimensionDataRowSerializer.prototype.deserialize = /**
     * @param {?} jobj
     * @return {?}
     */
    function (jobj) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        var dataRow = new DimensionDataRow();
        dataRow.id = jobj['id'];
        if (jobj['dimensionDataItems']) {
            dataRow.dimensionDataItems = new Array();
            jobj['dimensionDataItems'].forEach((/**
             * @param {?} dimensionDataItemJobj
             * @return {?}
             */
            function (dimensionDataItemJobj) {
                dataRow.dimensionDataItems.push(Object.assign(new DataItem(), dimensionDataItemJobj));
            }));
        }
        if (jobj['paramDataItems']) {
            dataRow.paramDataItems = new Array();
            jobj['paramDataItems'].forEach((/**
             * @param {?} paramDataItemsJobj
             * @return {?}
             */
            function (paramDataItemsJobj) {
                dataRow.dimensionDataItems.push(Object.assign(new DataItem(), paramDataItemsJobj));
            }));
        }
        return dataRow;
    };
    return DimensionDataRowSerializer;
}());
export { DimensionDataRowSerializer };
