// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/param-data-serializer/param-data-vo-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/param-data-serializer/param-data-vo-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { UbccParamDataVo } from '../../param-data-vo/ubcc-param-data-vo';
import { SerializerUtils } from '../../utils/serializer-utils';
var ParamDataVoSerializer = /** @class */ (function () {
    function ParamDataVoSerializer() {
    }
    /**
     * @param {?} paramData
     * @return {?}
     */
    ParamDataVoSerializer.prototype.serialize = /**
     * @param {?} paramData
     * @return {?}
     */
    function (paramData) {
        /** @type {?} */
        var jobj = new Object();
        if (!paramData) {
            return jobj;
        }
        jobj['templateId'] = paramData.templateId;
        jobj['bizTypeId'] = paramData.bizTypeId;
        if (paramData.data) {
            jobj['data'] = [];
            paramData.data.forEach((/**
             * @param {?} dataRow
             * @return {?}
             */
            function (dataRow) {
                /** @type {?} */
                var serializer = SerializerUtils.getUbccParamDataRowSerializerFactory().get(dataRow.getKind());
                jobj['data'].push(serializer.serialize(dataRow));
            }));
        }
        return jobj;
    };
    /**
     * @param {?} jobj
     * @return {?}
     */
    ParamDataVoSerializer.prototype.deserialize = /**
     * @param {?} jobj
     * @return {?}
     */
    function (jobj) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        var paramData = new UbccParamDataVo();
        paramData.templateId = jobj['templateId'];
        paramData.bizTypeId = jobj['bizTypeId'];
        if (jobj['data']) {
            paramData.data = new Array();
            jobj['data'].forEach((/**
             * @param {?} dataRowJobj
             * @return {?}
             */
            function (dataRowJobj) {
                /** @type {?} */
                var serializer = SerializerUtils.getUbccParamDataRowSerializerFactory().get(dataRowJobj['kind']);
                paramData.data.push(serializer.deserialize(dataRowJobj));
            }));
        }
        return paramData;
    };
    return ParamDataVoSerializer;
}());
export { ParamDataVoSerializer };
