// @farris/kendo-binding/esm5/lib/directives/discussion-group/farris_discussion_editor_binding.directive.js
import { Directive, Injector, Input, HostListener } from '@angular/core';
import { DiscussionEditorComponent } from '@farris/discussion-group';
import { ViewModel } from '@farris/devkit';
import { of, EMPTY } from 'rxjs';
import { tap, catchError } from 'rxjs/operators';
var KEY_UISTATE_REPLY_MESSAGE = 'DISCUSSION_REPLY_MESSAGE_INFO';
var KEY_MESSAGE_ON_COMMENT_ADD = 'onCommentAdd';
var FarrisDiscussionEditorBindingDirective = /** @class */ (function () {
    function FarrisDiscussionEditorBindingDirective(injector, discussionEditorComponent, viewModel) {
        this.injector = injector;
        this.discussionEditorComponent = discussionEditorComponent;
        this.viewModel = viewModel;
    }
    FarrisDiscussionEditorBindingDirective.prototype.ngOnInit = function () {
        var _this = this;
        this.discussionEditorComponent.replyUser = this.viewModel.frameContext.root.viewModel.uiState[KEY_UISTATE_REPLY_MESSAGE] || {};
        this.viewModel.frameContext.root.viewModel.uiState.changes.subscribe(function (change) {
            if (change.field === KEY_UISTATE_REPLY_MESSAGE) {
                _this.discussionEditorComponent.replyUser = _this.viewModel.frameContext.root.viewModel.uiState[KEY_UISTATE_REPLY_MESSAGE] || {};
            }
        });
        if (this.queryFrequentAtUsersCommand) {
            this.queryFrequentAtUsers();
        }
        else {
            this.queryAtUsers();
        }
        this.queryAllOrgs();
    };
    /**
     * 查询所有部门数据
     */
    FarrisDiscussionEditorBindingDirective.prototype.queryAllOrgs = function () {
        var _this = this;
        this.execute(this.queryAllOrgsCommand).subscribe(function (result) {
            if (!result) {
                return;
            }
            _this.discussionEditorComponent.sectionData = result;
        });
    };
    /**
     * 获取@用户
     */
    FarrisDiscussionEditorBindingDirective.prototype.queryAtUsers = function () {
        var _this = this;
        this.execute(this.userQueryCommand).subscribe(function (result) {
            if (!result) {
                return;
            }
            var _a = result.users, users = _a === void 0 ? [] : _a;
            _this.discussionEditorComponent.personnels = users;
        });
    };
    /**
     * 获取常用@用户
     */
    FarrisDiscussionEditorBindingDirective.prototype.queryFrequentAtUsers = function () {
        var _this = this;
        this.execute(this.queryFrequentAtUsersCommand).subscribe(function (result) {
            if (!result) {
                return;
            }
            _this.discussionEditorComponent.personnels = result && result.users || [];
        });
    };
    /**
     * 提交或取消评论
     * @param event event
     */
    FarrisDiscussionEditorBindingDirective.prototype.pageChangedHandler = function (event) {
        var _this = this;
        var _a = event || {}, _b = _a.msgInfo, msgInfo = _b === void 0 ? 0 : _b, _c = _a.text, text = _c === void 0 ? '' : _c, _d = _a.visibility, visibility = _d === void 0 ? 'ALL' : _d, _e = _a.parentId, parentId = _e === void 0 ? null : _e;
        if (msgInfo === 1) {
            this.execute(this.addCommentCommand, { text: text, parentId: parentId, visibility: visibility }).pipe(tap(function () {
                _this.viewModel.frameContext.appContext.messagePipe.next(KEY_MESSAGE_ON_COMMENT_ADD);
            }), catchError(function (e) { return EMPTY; })).subscribe();
        }
        else {
            // 取消的时候同时会将回复用户清空
            this.viewModel.frameContext.root.viewModel.uiState[KEY_UISTATE_REPLY_MESSAGE] = {};
            this.discussionEditorComponent.replyUser = {};
        }
    };
    /**
     * 执行命令
     * @param commandName 命令名称
     */
    FarrisDiscussionEditorBindingDirective.prototype.execute = function (commandName, params) {
        var _this = this;
        if (!commandName || commandName === '' || commandName === 'undefined') {
            return of(null);
        }
        if (typeof params === 'undefined') {
            params = {};
        }
        var paths = commandName.split('.');
        var func = null;
        paths.forEach(function (path) {
            func = func && func[path] || _this[path];
        });
        if (typeof func === 'function') {
            return func(params);
        }
        return of(null);
    };
    FarrisDiscussionEditorBindingDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-discussion-editor-binding]'
                },] }
    ];
    /** @nocollapse */
    FarrisDiscussionEditorBindingDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: DiscussionEditorComponent },
        { type: ViewModel }
    ]; };
    FarrisDiscussionEditorBindingDirective.propDecorators = {
        userQueryCommand: [{ type: Input, args: ["userQueryCommand",] }],
        addCommentCommand: [{ type: Input, args: ["addCommentCommand",] }],
        queryAllOrgsCommand: [{ type: Input, args: ["queryAllOrgsCommand",] }],
        queryFrequentAtUsersCommand: [{ type: Input, args: ["queryFrequentAtUsersCommand",] }],
        pageChangedHandler: [{ type: HostListener, args: ['valueChange', ['$event'],] }]
    };
    return FarrisDiscussionEditorBindingDirective;
}());
export { FarrisDiscussionEditorBindingDirective };
