// @farris/ui-enum-editor/esm2015/lib/enum-editor.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InputGroupModule } from '@farris/ui-input-group';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { MessagerModule } from '@farris/ui-messager';
import { LocaleModule } from '@farris/ui-locale';
import { NgModule } from '@angular/core';
import { EnumEditorInputComponent } from './enum-editor-input.component';
import { EnumEditorComponent } from './enum-editor.component';
import { EnumEditorService } from './enum-editor.service';
import { FarrisCommonModule } from '@farris/ui-common';
import { PerfectScrollbarModule } from '@farris/ui-perfect-scrollbar';
import { LookupModule } from '@farris/ui-lookup';
import { EnumEditorFooterComponent } from './enum-editor-footer.component';
export class EnumEditorModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: EnumEditorModule,
            providers: [
                EnumEditorService
            ]
        };
    }
}
EnumEditorModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    EnumEditorComponent,
                    EnumEditorInputComponent,
                    EnumEditorFooterComponent
                ],
                imports: [
                    CommonModule,
                    FormsModule,
                    InputGroupModule,
                    FarrisCommonModule,
                    MessagerModule.forRoot(),
                    PerfectScrollbarModule,
                    LookupModule,
                    LocaleModule
                ],
                entryComponents: [
                    EnumEditorComponent
                ],
                exports: [EnumEditorComponent, EnumEditorInputComponent]
            },] }
];
