// @farris/ui-enum-editor/esm2015/lib/enum-editor-input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InputGroupComponent } from '@farris/ui-input-group';
import { EnumEditorService, EnumOutType } from './enum-editor.service';
import { Component, Injector, ElementRef, Input, EventEmitter, Output, forwardRef, ViewChild, ChangeDetectorRef } from '@angular/core';
import { LookupComponent } from '@farris/ui-lookup';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/** @type {?} */
export const ENUM_EDITOR_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef((/**
     * @return {?}
     */
    () => EnumEditorInputComponent)),
    multi: true
};
export class EnumEditorInputComponent extends LookupComponent {
    /**
     * @param {?} injector
     * @param {?} ees
     * @param {?} el
     */
    constructor(injector, ees, el) {
        super(injector, el);
        this.injector = injector;
        this.ees = ees;
        this.el = el;
        this.textField = 'name';
        this.valueField = 'value';
        this.outType = EnumOutType.array;
        this.enableClear = true;
        this.valueChange = new EventEmitter();
        this.groupIcon = '<i class="f-icon f-icon-lookup"></i>';
        this.cd = null;
        this.ees.dataChanged.subscribe((/**
         * @param {?} str
         * @return {?}
         */
        str => {
            this.updateData(str);
        }));
        this.cd = this.injector.get(ChangeDetectorRef, null);
    }
    /**
     * @private
     * @param {?} str
     * @return {?}
     */
    updateData(str) {
        this.value = str;
        if (this.cd) {
            this.cd.detectChanges();
        }
        this.valueChange.emit(str);
        this.onModelChange(str);
        this.onModelTouched(str);
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    showDialog() {
        /** @type {?} */
        const enumData = this.convertValue2JSON();
        this.ees.showDialog(enumData, {
            textField: this.textField,
            valueField: this.valueField,
            outType: this.outType
        });
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClear($event) {
        this.updateData('');
    }
    /**
     * @param {?} val
     * @return {?}
     */
    writeValue(val) {
        if (val && typeof val === 'object') {
            this.value = JSON.stringify(val);
        }
        else {
            this.value = val || '';
        }
    }
    /**
     * @private
     * @return {?}
     */
    convertValue2JSON() {
        return this.ees.toJSON(this.value);
    }
}
EnumEditorInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'enum-editor-input',
                template: `
    <input-group #ig style="width: 100%"
        [groupText]="groupIcon"
        (clear)="onClear($event)"
        (clickHandle)="showDialog()"
        [readonly] = "readonly"
        [editable] = "editable"
        [enableClear] = "enableClear"
        [(value)]="value">
    </input-group>
    `,
                providers: [
                    ENUM_EDITOR_VALUE_ACCESSOR,
                    EnumEditorService
                ]
            }] }
];
/** @nocollapse */
EnumEditorInputComponent.ctorParameters = () => [
    { type: Injector },
    { type: EnumEditorService },
    { type: ElementRef }
];
EnumEditorInputComponent.propDecorators = {
    value: [{ type: Input }],
    textField: [{ type: Input }],
    valueField: [{ type: Input }],
    outType: [{ type: Input }],
    enableClear: [{ type: Input }],
    valueChange: [{ type: Output }],
    inputGroupRef: [{ type: ViewChild, args: ['ig',] }]
};
if (false) {
    /** @type {?} */
    EnumEditorInputComponent.prototype.value;
    /** @type {?} */
    EnumEditorInputComponent.prototype.textField;
    /** @type {?} */
    EnumEditorInputComponent.prototype.valueField;
    /** @type {?} */
    EnumEditorInputComponent.prototype.outType;
    /** @type {?} */
    EnumEditorInputComponent.prototype.enableClear;
    /** @type {?} */
    EnumEditorInputComponent.prototype.valueChange;
    /** @type {?} */
    EnumEditorInputComponent.prototype.inputGroupRef;
    /** @type {?} */
    EnumEditorInputComponent.prototype.groupIcon;
    /** @type {?} */
    EnumEditorInputComponent.prototype.cd;
    /** @type {?} */
    EnumEditorInputComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    EnumEditorInputComponent.prototype.ees;
    /** @type {?} */
    EnumEditorInputComponent.prototype.el;
}
