// @farris/ui-enum-editor/esm2015/lib/enum-editor.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { LocaleService } from '@farris/ui-locale';
import { BsModalService } from '@farris/ui-modal';
import { Injectable, Injector, ComponentFactoryResolver } from '@angular/core';
import { Subject } from 'rxjs';
import { EnumEditorComponent } from './enum-editor.component';
import { cloneDeep } from 'lodash-es';
import * as i0 from "@angular/core";
import * as i1 from "@farris/ui-modal";
import * as i2 from "@farris/ui-locale";
/** @enum {number} */
const EnumOutType = {
    'array': 0,
    'object': 1,
};
export { EnumOutType };
EnumOutType[EnumOutType['array']] = 'array';
EnumOutType[EnumOutType['object']] = 'object';
export class EnumEditorService {
    /**
     * @param {?} injector
     * @param {?} cfr
     * @param {?} modalService
     * @param {?} localeService
     */
    constructor(injector, cfr, modalService, localeService) {
        this.injector = injector;
        this.cfr = cfr;
        this.modalService = modalService;
        this.localeService = localeService;
        this.enumEditorRef = null;
        this.dlgRef = null;
        this.originalData = [];
        this.textField = 'name';
        this.valueField = 'value';
        this.outType = EnumOutType.array;
        this.dataChanged = new Subject();
    }
    /**
     * @param {?=} enumData
     * @param {?=} options
     * @return {?}
     */
    showDialog(enumData = [], options = {}) {
        enumData = this.toJSON(enumData);
        /** @type {?} */
        const enumEditorFactory = this.cfr.resolveComponentFactory(EnumEditorComponent);
        this.enumEditorRef = enumEditorFactory.create(this.injector);
        this.enumEditorRef.instance.data = enumData;
        this.originalData = cloneDeep(enumData);
        this.textField = options['textField'] || 'name';
        this.valueField = options['valueField'] || 'value';
        this.outType = options['outType'] || EnumOutType.array;
        options['showSortBtns'] = this.outType === EnumOutType.array;
        this.dlgRef = this.modalService.show(this.enumEditorRef, {
            width: 800, height: 500,
            title: this.localeService.getValue('enumEditor.title'), enableScroll: false,
            minHeight: 398, minWidth: 798, iconCls: 'f-icon f-icon-top_developmenttool',
            buttons: [
                {
                    text: this.localeService.getValue('enumEditor.cancelButton'),
                    cls: 'btn btn-outline-secondary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        this.cancel();
                    })
                },
                {
                    text: this.localeService.getValue('enumEditor.okButton'),
                    cls: 'btn btn-primary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        this.save();
                    })
                }
            ],
            initialState: options,
            dialogFooterStyles: { background: '#F4F6F9' },
            buttonAlign: 'right',
            closed: (/**
             * @param {?} isCloseButtonClick
             * @return {?}
             */
            (isCloseButtonClick) => {
                if (isCloseButtonClick) {
                    this.cancel();
                }
            })
        });
        this.enumEditorRef.instance.height = this.dlgRef.dialog.instance.getContainerSize().height;
        this.dlgRef.dialog.instance.resized.subscribe((/**
         * @param {?} size
         * @return {?}
         */
        size => {
            this.enumEditorRef.instance.height = size.containerHeight;
        }));
        this.enumEditorRef.changeDetectorRef.detectChanges();
        this.dlgRef.dialog.changeDetectorRef.reattach();
        // this.openDialog.emit();
    }
    /**
     * @return {?}
     */
    cancel() {
        this._dataChanged(this.originalData);
        this.dlgRef.close();
    }
    /**
     * @return {?}
     */
    save() {
        /** @type {?} */
        const enumData = this.enumEditorRef.instance.data;
        this._dataChanged(enumData);
        this.dlgRef.close();
    }
    /**
     * @private
     * @param {?} enumData
     * @return {?}
     */
    _dataChanged(enumData) {
        /** @type {?} */
        const str = this.toString(enumData);
        this.dataChanged.next(str);
    }
    /**
     * @param {?} enumData
     * @return {?}
     */
    toString(enumData) {
        if (enumData && enumData.length) {
            if (this.outType === EnumOutType.array) {
                return JSON.stringify(enumData);
            }
            else {
                /** @type {?} */
                const obj = enumData.reduce((/**
                 * @param {?} r
                 * @param {?} c
                 * @return {?}
                 */
                (r, c) => {
                    r[c[this.valueField]] = c[this.textField];
                    return r;
                }), {});
                return JSON.stringify(obj);
            }
        }
        return '';
    }
    /**
     * @param {?} val
     * @return {?}
     */
    toJSON(val) {
        /** @type {?} */
        let enumData = [];
        if (val) {
            if (typeof val === 'string') {
                try {
                    enumData = JSON.parse(val);
                    if (this.outType === EnumOutType.object) {
                        enumData = this.convertObject2Array(enumData);
                    }
                }
                catch (e) {
                    enumData = [];
                    // console.warn(e);
                }
            }
            else {
                if (Array.isArray(val)) {
                    enumData = val;
                }
                else {
                    enumData = this.convertObject2Array(val);
                }
            }
        }
        return enumData;
    }
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    convertObject2Array(obj) {
        /** @type {?} */
        const _enumData = [];
        Object.keys(obj).forEach((/**
         * @param {?} n
         * @return {?}
         */
        n => {
            _enumData.push({ [this.valueField]: n, [this.textField]: obj[n] });
        }));
        return _enumData;
    }
}
EnumEditorService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
EnumEditorService.ctorParameters = () => [
    { type: Injector },
    { type: ComponentFactoryResolver },
    { type: BsModalService },
    { type: LocaleService }
];
/** @nocollapse */ EnumEditorService.ngInjectableDef = i0.defineInjectable({ factory: function EnumEditorService_Factory() { return new EnumEditorService(i0.inject(i0.INJECTOR), i0.inject(i0.ComponentFactoryResolver), i0.inject(i1.BsModalService), i0.inject(i2.LocaleService)); }, token: EnumEditorService, providedIn: "root" });
if (false) {
    /** @type {?} */
    EnumEditorService.prototype.enumEditorRef;
    /** @type {?} */
    EnumEditorService.prototype.dlgRef;
    /** @type {?} */
    EnumEditorService.prototype.originalData;
    /** @type {?} */
    EnumEditorService.prototype.textField;
    /** @type {?} */
    EnumEditorService.prototype.valueField;
    /** @type {?} */
    EnumEditorService.prototype.outType;
    /** @type {?} */
    EnumEditorService.prototype.dataChanged;
    /**
     * @type {?}
     * @private
     */
    EnumEditorService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    EnumEditorService.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    EnumEditorService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    EnumEditorService.prototype.localeService;
}
