// @farris/bef/esm5/lib/utils/back_end_message.util.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/back_end_message.util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { AppContext, BackEndMessage, BACK_END_MESSAGE_HANDLER_TOKEN, NAMESPACE } from "@farris/devkit";
var BackEndMessageUtil = /** @class */ (function () {
    function BackEndMessageUtil() {
    }
    /**
     * @param {?} messages
     * @param {?} injector
     * @param {?=} options
     * @return {?}
     */
    BackEndMessageUtil.handleMessage = /**
     * @param {?} messages
     * @param {?} injector
     * @param {?=} options
     * @return {?}
     */
    function (messages, injector, options) {
        /** @type {?} */
        var namespace = injector.get(NAMESPACE, null);
        /** @type {?} */
        var appContext = injector.get(AppContext, null);
        /** @type {?} */
        var context = {
            ns: namespace,
            appContext: appContext
        };
        /** @type {?} */
        var bizMessages = this.buildBackEndMessages(messages);
        /** @type {?} */
        var backEndMessage = new BackEndMessage.Message(bizMessages, context);
        /** @type {?} */
        var backEndMessageHandler = injector.get(BACK_END_MESSAGE_HANDLER_TOKEN, null);
        if (backEndMessageHandler) {
            backEndMessageHandler.handle(backEndMessage, options);
        }
    };
    /**
     * @param {?} injector
     * @return {?}
     */
    BackEndMessageUtil.isBackEndMessageHandlerExist = /**
     * @param {?} injector
     * @return {?}
     */
    function (injector) {
        /** @type {?} */
        var backEndMessageHandler = injector.get(BACK_END_MESSAGE_HANDLER_TOKEN, null);
        return backEndMessageHandler && true || false;
    };
    /**
     * @param {?} messages
     * @return {?}
     */
    BackEndMessageUtil.getFormlessMessages = /**
     * @param {?} messages
     * @return {?}
     */
    function (messages) {
        /** @type {?} */
        var bizMessages = this.buildBackEndMessages(messages) || [];
        return bizMessages.filter((/**
         * @param {?} message
         * @return {?}
         */
        function (message) {
            return !message.location || !message.location.columns || message.location.columns.length < 1 || !message.location.nodeCode;
        }));
    };
    /**
     * @param {?} messages
     * @return {?}
     */
    BackEndMessageUtil.buildBackEndMessages = /**
     * @param {?} messages
     * @return {?}
     */
    function (messages) {
        if (!messages || messages.length < 1 || !Array.isArray(messages)) {
            return null;
        }
        /** @type {?} */
        var bizMessages = [];
        messages.forEach((/**
         * @param {?} message
         * @return {?}
         */
        function (message) {
            /** @type {?} */
            var bizMessage = {
                level: message.level,
                message: message.message
            };
            if (message.hasOwnProperty('location') && message.location) {
                bizMessage.location = {
                    nodeCode: message.location.nodeCode,
                    rows: message.location.dataIds,
                    columns: message.location.columnNames
                };
            }
            bizMessages.push(bizMessage);
        }));
        return bizMessages;
    };
    return BackEndMessageUtil;
}());
export { BackEndMessageUtil };
