// @gsp-cmp/chgdr/esm5/lib/chgdr.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/chgdr.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule, registerLocaleData } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { TranslateModule } from '@ngx-translate/core';
import { SplitterModule } from "@farris/ui-splitter";
import { AngularDraggableModule } from '@farris/ui-draggable';
import { DatagridModule } from "@farris/ui-datagrid";
import { TreeTableModule } from '@farris/ui-treetable';
import { FarrisSectionModule } from "@farris/ui-section";
import { MessagerModule, MessagerService } from '@farris/ui-messager';
import { NotifyModule, NotifyService } from '@farris/ui-notify';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { FarrisDatePickerModule } from '@farris/ui-datepicker';
import { HttpService, SessionService, CacheService } from '@ecp-caf/caf-common';
import { MetadataRTSelectModule, GSPMetadataRTServiceModule } from '@gsp-lcm/metadatart-selector';
import { ChgdrDataViewerComponent } from './chgdr-data-viewer/chgdr-data-viewer.component';
import zh from '@angular/common/locales/zh';
import { ChgdrListComponent } from './chgdr-list/chgdr-list.component';
import { ChgdrRowBaseinfoFormComponent } from './chgdr-row-baseinfo-form/chgdr-row-baseinfo-form.component';
import { ChgdrRowGridComponent } from './chgdr-row-grid/chgdr-row-grid.component';
import { ChgdrListUiService } from './service/chgdr-list-ui.service';
import { ChgdrListDialogContentComponent } from './chgdr-list-dialog-content/chgdr-list-dialog-content.component';
import { ChgdrHeaderDatagridComponent } from './chgdr-header-datagrid/chgdr-header-datagrid.component';
import { ChgdrMetadataService } from './service/chgdr-metadata.service';
import { ChgdrConfigService } from './service/chgdr-config.service';
import { ChgdrInnerService } from './service/chgdr-inner.service';
import { ChgdrService } from './service/chgdr.service';
import { I18nService } from './service/i18n-service';
registerLocaleData(zh);
var ChgdrModule = /** @class */ (function () {
    function ChgdrModule() {
    }
    ChgdrModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [ChgdrDataViewerComponent, ChgdrListComponent, ChgdrRowBaseinfoFormComponent, ChgdrRowGridComponent, ChgdrListDialogContentComponent, ChgdrHeaderDatagridComponent],
                    imports: [
                        CommonModule,
                        FormsModule,
                        TranslateModule.forRoot(),
                        SplitterModule,
                        AngularDraggableModule,
                        DatagridModule,
                        TreeTableModule,
                        FarrisSectionModule,
                        MessagerModule.forRoot(),
                        NotifyModule,
                        FarrisDialogModule,
                        FarrisDatePickerModule,
                        MetadataRTSelectModule,
                        GSPMetadataRTServiceModule.forRoot('')
                    ],
                    providers: [
                        MessagerService,
                        NotifyService,
                        HttpService,
                        SessionService,
                        CacheService,
                        ChgdrListUiService,
                        ChgdrMetadataService,
                        ChgdrConfigService,
                        ChgdrInnerService,
                        ChgdrService,
                        I18nService
                    ],
                    entryComponents: [
                        ChgdrDataViewerComponent,
                        ChgdrListComponent,
                        ChgdrListDialogContentComponent
                    ],
                    exports: [ChgdrDataViewerComponent, ChgdrListComponent, ChgdrHeaderDatagridComponent]
                },] }
    ];
    return ChgdrModule;
}());
export { ChgdrModule };
