// @gsp-cmp/querysolution/esm5/lib/query-solution.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/query-solution.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { QuerySolutionComponent } from './query-solution.component';
import { SolutionManangerComponent } from './solutionmanager/solutionmanager.component';
import { SaveAsDialogComponent } from './saveasdialog/saveasdialog.component';
import { FormsModule } from '@angular/forms';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { LookupModule } from '@farris/ui-lookup';
import { MultiSelectModule } from '@farris/ui-multi-select';
import { FieldGroupModule } from '@farris/ui-field-group';
import { InputGroupModule } from '@farris/ui-input-group';
import { FDropdownDirectiveTypeModule } from '@farris/ui-dropdown';
import { NotifyModule } from '@farris/ui-notify';
import { LoadingModule } from '@farris/ui-loading';
import { HttpClientModule } from '@angular/common/http';
import { SolutionService, QueryConditionModule } from '@farris/component-querycondition';
import { LocaleModule } from '@farris/ui-locale';
import { DatagridModule } from '@farris/ui-datagrid';
import { NumberHelperModule } from '@farris/ui-common/number';
import { DateHelperModule } from '@farris/ui-common/date';
import { FarrisTooltipModule } from '@farris/ui-tooltip';
import { SummaryConditionComponent } from './summary-condition/summary-condition.component';
import { SummaryListComponent } from './summary-list/summary-list.component';
import { FarrisFormsModule } from '@farris/ui-forms';
import { OrganizationSelectorModule } from '@farris/ui-organization-selector';
import { LanguageTextboxModule } from '@farris/ui-language-textbox';
var QuerySolutionModule = /** @class */ (function () {
    function QuerySolutionModule() {
    }
    QuerySolutionModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        FarrisDialogModule,
                        LookupModule,
                        MultiSelectModule,
                        FieldGroupModule,
                        InputGroupModule,
                        FDropdownDirectiveTypeModule,
                        NotifyModule.forRoot(),
                        LoadingModule.forRoot(),
                        QueryConditionModule,
                        HttpClientModule,
                        LocaleModule,
                        DatagridModule,
                        NumberHelperModule,
                        DateHelperModule,
                        FarrisTooltipModule,
                        FarrisFormsModule,
                        OrganizationSelectorModule,
                        LanguageTextboxModule
                    ],
                    declarations: [
                        QuerySolutionComponent,
                        SolutionManangerComponent,
                        SaveAsDialogComponent,
                        SummaryConditionComponent,
                        SummaryListComponent,
                    ],
                    entryComponents: [
                        SolutionManangerComponent,
                        SaveAsDialogComponent
                    ],
                    providers: [SolutionService],
                    exports: [
                        QuerySolutionComponent,
                        SolutionManangerComponent,
                        SaveAsDialogComponent
                    ]
                },] }
    ];
    return QuerySolutionModule;
}());
export { QuerySolutionModule };
