// @gsp-bef/gsp-be-metadata/esm2015/lib/biz-entity/src/Enum/BEOperationType.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
* BE 操作类型
* @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {number} */
const BEOperationType = {
    /**
     * 自定义操作
     */
    BizMgrAction: 0,
    /**
     * 业务实体操作
     */
    BizAction: 1,
    /**
     * 数据完整性检查
     */
    Validation: 2,
    /**
     * 自动计算
     */
    Determination: 3,
    /**
     * 查询
     */
    Query: 4,
    /**
     * 事件订阅
     */
    Subscription: 5,
    /**
     * 初始化
     */
    Initiation: 6,
    /**
     * 系统服务
     */
    BasicService: 7,
};
BEOperationType[BEOperationType.BizMgrAction] = 'BizMgrAction';
BEOperationType[BEOperationType.BizAction] = 'BizAction';
BEOperationType[BEOperationType.Validation] = 'Validation';
BEOperationType[BEOperationType.Determination] = 'Determination';
BEOperationType[BEOperationType.Query] = 'Query';
BEOperationType[BEOperationType.Subscription] = 'Subscription';
BEOperationType[BEOperationType.Initiation] = 'Initiation';
BEOperationType[BEOperationType.BasicService] = 'BasicService';
export { BEOperationType };
