// @edp-pmf/grapheditor/dist/cjs/lib/action/Action.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Action = void 0;
var tslib_1 = require("tslib");
var mxgraph_ts_1 = require("@edp-pmf/mxgraph-ts");
var PmfEvents_1 = require("../util/PmfEvents");
var KeyboardMap_1 = require("../util/KeyboardMap");
/**
 * 动作类
 */
var Action = /** @class */ (function (_super) {
    tslib_1.__extends(Action, _super);
    function Action(key, editorUi, enabled) {
        var _this = _super.call(this) || this;
        /**
         * 是否启用
         */
        _this.enabled = true;
        /**
         * 是否是切换动作
         */
        _this.toggleAction = false;
        _this.key = key;
        _this.editorUi = editorUi;
        _this.enabled = (enabled != null) ? enabled : true;
        return _this;
    }
    /**
     * 获取启用状态
     */
    Action.prototype.isEnabled = function () {
        return this.enabled;
    };
    /**
     * 获取是否为切换动作
     */
    Action.prototype.isToggleAction = function () {
        return this.toggleAction;
    };
    /**
     * 是否被选中
     */
    Action.prototype.isSelected = function () {
        return false;
    };
    /**
     * 设置启用状态，状态修改时发射状态变更事件
     */
    Action.prototype.setEnabled = function (value) {
        var origin = this.enabled;
        this.enabled = value;
        if (origin != value) {
            this.fireEvent(new mxgraph_ts_1.mxEventObject(PmfEvents_1.PmfEvents.StateChanged));
        }
    };
    /**
     * 获取快捷键文本
     */
    Action.prototype.getShortcut = function () {
        if (this.shortcutKeyCode == null)
            return null;
        var shortcut = KeyboardMap_1.KEYBOARD_MAP[this.shortcutKeyCode];
        if (this.shortcutAltKey) {
            shortcut = "Alt+" + shortcut;
        }
        if (this.shortcutShiftKey) {
            shortcut = "Shift+" + shortcut;
        }
        if (this.shortcutControlKey) {
            shortcut = Action.ctrlKey + "+" + shortcut;
        }
        return shortcut;
    };
    /**
     * 设置是否为切换动作
     */
    Action.prototype.setToggleAction = function (value) {
        this.toggleAction = value;
    };
    /**
     * 执行动作
     */
    Action.prototype.execute = function (evt) {
    };
    Action.ctrlKey = (mxgraph_ts_1.mxClient.IS_MAC) ? 'Cmd' : 'Ctrl';
    return Action;
}(mxgraph_ts_1.mxEventSource));
exports.Action = Action;
