// @edp-pmf/grapheditor/dist/cjs/lib/action/ActionKeys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionKeys = void 0;
/**
 * 动作Key列表
 */
var ActionKeys;
(function (ActionKeys) {
    //temp ="pmf_temp",
    // File
    ActionKeys["pmf_save"] = "pmf_save";
    ActionKeys["pmf_saveAs"] = "pmf_saveAs";
    ActionKeys["wf_sync"] = "wf_sync";
    ActionKeys["wf_note"] = "wf_note";
    ActionKeys["wf_edit"] = "wf_edit";
    ActionKeys["pmf_export"] = "pmf_export";
    //Edit
    ActionKeys["pmf_undo"] = "pmf_undo";
    ActionKeys["pmf_redo"] = "pmf_redo";
    ActionKeys["pmf_cut"] = "pmf_cut";
    ActionKeys["pmf_copy"] = "pmf_copy";
    ActionKeys["pmf_paste"] = "pmf_paste";
    ActionKeys["pmf_pasteHere"] = "pmf_pasteHere";
    ActionKeys["pmf_delete"] = "pmf_delete";
    ActionKeys["pmf_deleteAll"] = "pmf_deleteAll";
    ActionKeys["pmf_duplicate"] = "pmf_duplicate";
    ActionKeys["pmf_selectVertices"] = "pmf_selectVertices";
    ActionKeys["pmf_selectEdges"] = "pmf_selectEdges";
    ActionKeys["pmf_selectAll"] = "pmf_selectAll";
    ActionKeys["pmf_selectNone"] = "pmf_selectNone";
    ActionKeys["pmf_lockUnlock"] = "pmf_lockUnlock";
    //View
    ActionKeys["pmf_formatPanel"] = "pmf_formatPanel";
    ActionKeys["pmf_outline"] = "pmf_outline";
    ActionKeys["pmf_layers"] = "pmf_layers";
    ActionKeys["pmf_pageView"] = "pmf_pageView";
    ActionKeys["pmf_grid"] = "pmf_grid";
    ActionKeys["pmf_connectionArrows"] = "pmf_connectionArrows";
    ActionKeys["pmf_connectionPoints"] = "pmf_connectionPoints";
    ActionKeys["pmf_resetView"] = "pmf_resetView";
    ActionKeys["pmf_zoomIn"] = "pmf_zoomIn";
    ActionKeys["pmf_zoomOut"] = "pmf_zoomOut";
    ActionKeys["pmf_view_25"] = "pmf_view_25";
    ActionKeys["pmf_view_50"] = "pmf_view_50";
    ActionKeys["pmf_view_75"] = "pmf_view_75";
    ActionKeys["pmf_view_100"] = "pmf_view_100";
    ActionKeys["pmf_view_125"] = "pmf_view_125";
    ActionKeys["pmf_view_150"] = "pmf_view_150";
    ActionKeys["pmf_view_200"] = "pmf_view_200";
    ActionKeys["pmf_view_300"] = "pmf_view_300";
    ActionKeys["pmf_view_400"] = "pmf_view_400";
    ActionKeys["pmf_fitWindow"] = "pmf_fitWindow";
    ActionKeys["pmf_fitPage"] = "pmf_fitPage";
    ActionKeys["pmf_fitTwoPages"] = "pmf_fitTwoPages";
    ActionKeys["pmf_fitPageWidth"] = "pmf_fitPageWidth";
    ActionKeys["pmf_customZoom"] = "pmf_customZoom";
    // Arrange
    ActionKeys["pmf_toFront"] = "pmf_toFront";
    ActionKeys["pmf_toBack"] = "pmf_toBack";
    ActionKeys["pmf_leftAlign"] = "pmf_leftAlign";
    ActionKeys["pmf_center"] = "pmf_center";
    ActionKeys["pmf_rightAlign"] = "pmf_rightAlign";
    ActionKeys["pmf_topAlign"] = "pmf_topAlign";
    ActionKeys["pmf_middle"] = "pmf_middle";
    ActionKeys["pmf_bottomAlign"] = "pmf_bottomAlign";
    ActionKeys["pmf_horizontal"] = "pmf_horizontal";
    ActionKeys["pmf_vertical"] = "pmf_vertical";
    ActionKeys["pmf_clearWaypoints"] = "pmf_clearWaypoints";
    ActionKeys["pmf_autosize"] = "pmf_autosize";
    //Help
    ActionKeys["pmf_help"] = "pmf_help";
    ActionKeys["pmf_about"] = "pmf_about";
    ActionKeys["pmf_viewModelXml"] = "pmf_viewModelXml";
    ActionKeys["pmf_release"] = "pmf_release";
    ActionKeys["pmf_closeDesigner"] = "pmf_closeDesigner";
})(ActionKeys = exports.ActionKeys || (exports.ActionKeys = {}));
