// @progress/kendo-angular-dialog/dist/npm/window/draggable.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var kendo_draggable_1 = require("@telerik/kendo-draggable");
/**
 * @hidden
 */
var DraggableDirective = /** @class */ (function () {
    function DraggableDirective(elem, ngZone) {
        this.elem = elem;
        this.ngZone = ngZone;
        this.kendo = {
            drag: new core_1.EventEmitter(),
            press: new core_1.EventEmitter(),
            release: new core_1.EventEmitter()
        };
    }
    DraggableDirective.prototype.ngOnInit = function () {
        var _this = this;
        if (typeof document !== 'undefined') {
            this.draggable = new kendo_draggable_1.default({
                drag: function (e) { return _this.kendo.drag.next(e); },
                press: function (e) { return _this.kendo.press.next(e); },
                release: function (e) { return _this.kendo.release.next(e); }
            });
            this.ngZone.runOutsideAngular(function () { return _this.draggable.bindTo(_this.elem.nativeElement); });
        }
    };
    DraggableDirective.prototype.ngOnDestroy = function () {
        if (typeof document !== 'undefined') {
            this.draggable.destroy();
        }
    };
    Object.defineProperty(DraggableDirective.prototype, "nativeElement", {
        get: function () {
            return this.elem.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    DraggableDirective.decorators = [
        { type: core_1.Directive, args: [{
                    outputs: ['kendo.press', 'kendo.drag', 'kendo.release'],
                    selector: '[kendoWindowDraggable]'
                },] },
    ];
    /** @nocollapse */
    DraggableDirective.ctorParameters = function () { return [
        { type: core_1.ElementRef, },
        { type: core_1.NgZone, },
    ]; };
    return DraggableDirective;
}());
exports.DraggableDirective = DraggableDirective;
