// @edp-aif/runtime-api/esm5/lib/aif-manual-creation/service/aif-data.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from "@angular/core";
import { FormSettings, GeneralTextEntity } from "@edp-aif/common-api";
import { of } from "rxjs";
import { FormGetCreationRuleEntityExArgs, FormGetCreationRuleEntityExResult, FormCreateResult, FormUpdateResult } from '../form-args';
import { map, switchMap } from 'rxjs/operators';
import { Repository } from '@farris/devkit';
import { RpcCreationRuleExecutorResult } from '../../entity';
var AifDataService = /** @class */ (function () {
    function AifDataService(repository) {
        this.getCreationRuleEntityByIdUrl = "/api/runtime/aif/v1.0/form/getcreationruleentityex";
        this.getCreationRuleEntityByIdMethod = "GET";
        this.getCreationRuleUrl = "/api/runtime/aif/v1.0/form/getcreationruleentityexes";
        this.getCreationRuleMethod = "POST";
        this.executeCreationRuleUrl = "/api/runtime/aif/v1.0/form/executecreationrule";
        this.executeCreationRuleMethod = "POST";
        this.formCreateUrl = "/api/runtime/aif/v1.0/form/create";
        this.formCreateMethod = "POST";
        this.formUpdateUrl = "/api/runtime/aif/v1.0/form/update";
        this.formUpdateMethod = "POST";
        this.generalTextUrl = "/api/runtime/aif/v1.0/generaltext";
        this.generalTextMethod = "GET";
        this.repository = (/** @type {?} */ (repository));
    }
    /**
     * 根据来源 VO Id获取生单规则实体列表。
     *
     * @param sourceVoId 来源 VO Id。
     * @return 符合条件的生单规则实体列表。
     */
    /**
     * 根据来源 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId 来源 VO Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    AifDataService.prototype.getSlimCreationRulesBySourceVoId = /**
     * 根据来源 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId 来源 VO Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    function (sourceVoId) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        return this.getCreationRuleList(args).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res.ruleEntityExes;
        })));
    };
    /**
     * 根据目标 VO Id获取生单规则实体列表。
     *
     * @param targetVoId 目标 VO Id。
     * @return 符合条件的生单规则实体列表。
     */
    /**
     * 根据目标 VO Id获取生单规则实体列表。
     *
     * @param {?} targetVoId 目标 VO Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    AifDataService.prototype.getSlimCreationRulesByTargetVoId = /**
     * 根据目标 VO Id获取生单规则实体列表。
     *
     * @param {?} targetVoId 目标 VO Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    function (targetVoId) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.targetVoId = targetVoId;
        return this.getCreationRuleList(args).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res.ruleEntityExes;
        })));
    };
    /**
     * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
     *
     * @param sourceVoId 来源 VO Id。
     * @param targetVoId 目标 VO Id。
     * @return 符合条件的生单规则实体列表。
     */
    /**
     * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId 来源 VO Id。
     * @param {?} targetVoId 目标 VO Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    AifDataService.prototype.getSlimCreationRulesBySourceVoIdAndTargetVoId = /**
     * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId 来源 VO Id。
     * @param {?} targetVoId 目标 VO Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    function (sourceVoId, targetVoId) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        args.targetVoId = targetVoId;
        return this.getCreationRuleList(args).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res.ruleEntityExes;
        })));
    };
    /**
     * 根据来源 VO Id获取生单规则实体列表。
     *
     * @param sourceVoId     来源 VO Id。
     * @param bizFlowchartId 业务场景图Id。
     * @return 符合条件的生单规则实体列表。
     */
    /**
     * 根据来源 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId     来源 VO Id。
     * @param {?} bizFlowchartId 业务场景图Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    AifDataService.prototype.getSlimCreationRulesBySourceVoIdAndFlowchartId = /**
     * 根据来源 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId     来源 VO Id。
     * @param {?} bizFlowchartId 业务场景图Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    function (sourceVoId, bizFlowchartId) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        args.bizFlowchartId = bizFlowchartId;
        return this.getCreationRuleList(args).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res.ruleEntityExes;
        })));
    };
    /**
     * 根据目标 VO Id获取生单规则实体列表。
     *
     * @param targetVoId     目标 VO Id。
     * @param bizFlowchartId 业务场景图Id。
     * @return 符合条件的生单规则实体列表。
     */
    /**
     * 根据目标 VO Id获取生单规则实体列表。
     *
     * @param {?} targetVoId     目标 VO Id。
     * @param {?} bizFlowchartId 业务场景图Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    AifDataService.prototype.getSlimCreationRulesByTargetVoIdAndFlowchartId = /**
     * 根据目标 VO Id获取生单规则实体列表。
     *
     * @param {?} targetVoId     目标 VO Id。
     * @param {?} bizFlowchartId 业务场景图Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    function (targetVoId, bizFlowchartId) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.targetVoId = targetVoId;
        args.bizFlowchartId = bizFlowchartId;
        return this.getCreationRuleList(args).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res.ruleEntityExes;
        })));
    };
    /**
     * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
     *
     * @param sourceVoId     来源 VO Id。
     * @param targetVoId     目标 VO Id。
     * @param bizFlowchartId 业务场景图Id。
     * @return 符合条件的生单规则实体列表。
     */
    /**
     * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId     来源 VO Id。
     * @param {?} targetVoId     目标 VO Id。
     * @param {?} bizFlowchartId 业务场景图Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    AifDataService.prototype.getSlimCreationRulesBySourceVoIdAndTargetVoIdAndFlowchartId = /**
     * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId     来源 VO Id。
     * @param {?} targetVoId     目标 VO Id。
     * @param {?} bizFlowchartId 业务场景图Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    function (sourceVoId, targetVoId, bizFlowchartId) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        args.targetVoId = targetVoId;
        args.bizFlowchartId = bizFlowchartId;
        return this.getCreationRuleList(args).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res.ruleEntityExes;
        })));
    };
    /**
     * 获取生单规则实体
     */
    /**
     * 获取生单规则实体
     * @param {?} creationRuleId
     * @return {?}
     */
    AifDataService.prototype.getSlimCreationRuleEntityExById = /**
     * 获取生单规则实体
     * @param {?} creationRuleId
     * @return {?}
     */
    function (creationRuleId) {
        /** @type {?} */
        var getCreationRuleById$ = this.repository.proxy.request(this.getCreationRuleEntityByIdUrl + "/" + creationRuleId, this.getCreationRuleEntityByIdMethod, null);
        return getCreationRuleById$;
    };
    /**
     * 获取表单配置实体
     * @param formSettingsId
     */
    /**
     * 获取表单配置实体
     * @param {?} formSettingsId
     * @return {?}
     */
    AifDataService.prototype.getFormSettingsById = /**
     * 获取表单配置实体
     * @param {?} formSettingsId
     * @return {?}
     */
    function (formSettingsId) {
        if (formSettingsId == null || formSettingsId.trim() == "") {
            throw "生单规则为空";
        }
        /** @type {?} */
        var formSettings$ = this.repository.proxy.request(this.generalTextUrl + "/" + formSettingsId, this.generalTextMethod, null, { responseType: "text" });
        return formSettings$.pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var generateText = new GeneralTextEntity();
            generateText.LoadFromJson(data);
            /** @type {?} */
            var formSettings = new FormSettings();
            formSettings.LoadFromJson(generateText.text);
            return formSettings;
        })));
    };
    /**
     * 执行生单规则
     */
    /**
     * 执行生单规则
     * @param {?} args
     * @return {?}
     */
    AifDataService.prototype.executeCreationRule = /**
     * 执行生单规则
     * @param {?} args
     * @return {?}
     */
    function (args) {
        /** @type {?} */
        var body = args.ConvertToJson();
        /** @type {?} */
        var executeCreationRule$ = this.repository.proxy.request(this.executeCreationRuleUrl, this.executeCreationRuleMethod, null, {
            body: body,
            responseType: 'text'
        });
        return executeCreationRule$.pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var res = new RpcCreationRuleExecutorResult();
            res.LoadFromJson(data);
            return res;
        })));
    };
    /**
     * 创建单据
     */
    /**
     * 创建单据
     * @param {?} args
     * @return {?}
     */
    AifDataService.prototype.createEntityDatas = /**
     * 创建单据
     * @param {?} args
     * @return {?}
     */
    function (args) {
        var _this = this;
        /** @type {?} */
        var body = args.ConvertToJson();
        /** @type {?} */
        var options = {
            body: body,
            responseType: 'text'
        };
        /** @type {?} */
        var create$ = this.repository.proxy.request(this.formCreateUrl, this.formCreateMethod, null, options);
        return create$.pipe(switchMap((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            /** @type {?} */
            var result = new FormCreateResult();
            result.LoadFromJson(res);
            /** @type {?} */
            var entities = result.entityDataList.map((/**
             * @param {?} entityStr
             * @return {?}
             */
            function (entityStr) {
                /** @type {?} */
                var newEntity = _this.repository.buildEntity(JSON.parse(entityStr));
                _this.appendInitialData(newEntity, newEntity);
                return newEntity;
            }));
            /** @type {?} */
            var origin = _this.repository.entityCollection.getAllEntities();
            if (origin == null || origin.length == 0) {
                _this.repository.entityCollection.loadEntities(entities);
            }
            else {
                _this.repository.entityCollection.addEntities(entities);
            }
            return of(result);
        })));
    };
    /**
     * 修改单据
     */
    /**
     * 修改单据
     * @param {?} args
     * @return {?}
     */
    AifDataService.prototype.updateEntityDatas = /**
     * 修改单据
     * @param {?} args
     * @return {?}
     */
    function (args) {
        /** @type {?} */
        var body = args.ConvertToJson();
        /** @type {?} */
        var options = {
            body: body,
            responseType: "text"
        };
        /** @type {?} */
        var update$ = this.repository.proxy.request(this.formUpdateUrl, this.formUpdateMethod, null, options);
        return update$.pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var res = new FormUpdateResult();
            res.LoadFromJson(data);
            return res;
        })));
    };
    /**
     * @private
     * @param {?} args
     * @return {?}
     */
    AifDataService.prototype.getCreationRuleList = /**
     * @private
     * @param {?} args
     * @return {?}
     */
    function (args) {
        if (args == null) {
            throw "args不可为空";
        }
        if (typeof args == "string") {
            if (args == "") {
                throw "args不可为空";
            }
            /** @type {?} */
            var temp = new FormGetCreationRuleEntityExArgs();
            temp.LoadFromJson(args);
            args = temp;
        }
        /** @type {?} */
        var getCreationRule$ = this.repository.proxy.request(this.getCreationRuleUrl, this.getCreationRuleMethod, null, { body: args.ConvertToJson(), responseType: "text" });
        return getCreationRule$.pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var result = new FormGetCreationRuleEntityExResult();
            result.LoadFromJson(data);
            return result;
        })));
    };
    /**
 * 参考farris/bef EntityUtil实现
 * @param entity
 * @param initialData
 */
    /**
     * 参考farris/bef EntityUtil实现
     * @private
     * @param {?} entity
     * @param {?} initialData
     * @return {?}
     */
    AifDataService.prototype.appendInitialData = /**
     * 参考farris/bef EntityUtil实现
     * @private
     * @param {?} entity
     * @param {?} initialData
     * @return {?}
     */
    function (entity, initialData) {
        /** @type {?} */
        var data = Object.assign({}, initialData);
        delete data.id;
        delete data.parentID;
        entity['initialData'] = data;
    };
    AifDataService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AifDataService.ctorParameters = function () { return [
        { type: Repository }
    ]; };
    return AifDataService;
}());
export { AifDataService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.getCreationRuleEntityByIdUrl;
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.getCreationRuleEntityByIdMethod;
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.getCreationRuleUrl;
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.getCreationRuleMethod;
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.executeCreationRuleUrl;
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.executeCreationRuleMethod;
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.formCreateUrl;
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.formCreateMethod;
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.formUpdateUrl;
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.formUpdateMethod;
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.generalTextUrl;
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.generalTextMethod;
    /**
     * @type {?}
     * @private
     */
    AifDataService.prototype.repository;
}
