// @farris/devkit/esm5/lib/form/decorators.js
import { makePropDecorator, makeDecorator } from '../metadata/index';
/**
 * ----------------------------------------
 * NgValidateForm
 * ----------------------------------------
 */
/**
 * 子表单装饰器名称
 */
export var NG_VALIDATE_FORM = 'NgValidateForm';
/**
 * 表单装饰器工厂
 */
export function NgValidateForm(options) {
    var decoratorFactory = makeDecorator(NG_VALIDATE_FORM, function (obj) { return obj; });
    return decoratorFactory(options);
}
/**
 * ----------------------------------------
 * NgChildForm
 * ----------------------------------------
 */
/**
 * 子表单装饰器名称
 */
export var NG_CHILD_FORM = 'NgChildForm';
/**
 * 子表单装饰器工厂
 */
export var NgChildForm = makePropDecorator(NG_CHILD_FORM, function (obj) { return obj; });
/**
 * ----------------------------------------
 * NgChildFormArray
 * ----------------------------------------
 */
/**
 * 子表单数组装饰器名称
 * @type {string}
 */
export var NG_CHILD_FORM_ARRAY = 'NgChildFormArray';
/**
 * 子表单数组装饰器工厂
 */
export var NgChildFormArray = makePropDecorator(NG_CHILD_FORM_ARRAY, function (obj) { return obj; });
/**
 * ----------------------------------------
 * NgFormControl
 * ----------------------------------------
 */
/**
 * 表单控件装饰器名称
 * @type {string}
 */
export var NG_FORM_CONTROL = 'NgFormControl';
/**
 * 表单控件装饰器工厂
 */
export var NgFormControl = makePropDecorator(NG_FORM_CONTROL, function (obj) { return obj; });
