// @gsp-wf/task-action-impl/esm5/lib/task-action-impl/task-action-impl.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewChild, ViewContainerRef, Injector, ComponentFactoryResolver, Input, Output, EventEmitter, ElementRef, NgZone } from '@angular/core';
import { TaskExecuteComponent } from '@gsp-wf/wf-task-handler';
import { TaskEntity } from '@gsp-wf/task-impl-api';
import { Subject } from 'rxjs';
import { LoadingService } from '@farris/ui-loading';
import { MessagerService } from '@farris/ui-messager';
import { ActionOrigin } from '@gsp-wf/task-impl-api';
import { TaskActionImplService } from '../task-action-impl.service';
var TaskActionImplComponent = /** @class */ (function () {
    // 注意：修改构造方法需要重新打包task-main和task form
    function TaskActionImplComponent(injector, cfr, loadingService, msgService, taskService) {
        this.injector = injector;
        this.cfr = cfr;
        this.loadingService = loadingService;
        this.msgService = msgService;
        this.taskService = taskService;
        this.showButtons = true; //是否展示提交按钮。任务办理界面展示，任务列表不展示
        //是否展示提交按钮。任务办理界面展示，任务列表不展示
        this.batch = false;
        // 来源，从哪打开的
        this.submited = new EventEmitter();
        this.actionCode = new EventEmitter();
        this.actionCodeMap = new Map();
        this.index = 0;
        this.submitDisable = false; // 提交按钮是否禁用
        // 提交按钮是否禁用
        this.ifFocused = false; // 是否关注了此任务
        // 是否关注了此任务
        this.focus = this.i18nService("focus"); //关注此流程
        //关注此流程
        this.ifShowFocus = true; // 是否启用我的关注功能
        this.ngZone = this.injector.get(NgZone);
    }
    Object.defineProperty(TaskActionImplComponent.prototype, "iframe", {
        get: /**
         * @return {?}
         */
        function () {
            return this._iframe;
        },
        set: /**
         * @param {?} iframe
         * @return {?}
         */
        function (iframe) {
            this._iframe = iframe;
            if (this.iframe && this.componentRef) {
                this.componentRef.instance.iframe = this.iframe;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TaskActionImplComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        //表单触发预解析，重新加载组件
        ((/** @type {?} */ (window))).gspWfEventService = ((/** @type {?} */ (window))).gspWfEventService || {};
        ((/** @type {?} */ (window))).gspWfEventService.simulate = function (func) {
            _this.ngZone.run(function () {
                _this.actionSelect(_this.actions[0], 0).subscribe(function () { return func && typeof func === 'function' && func(); });
            });
        };
        // 是否启用我的关注功能
        /** @type {?} */
        var load = this.loadingService.show();
        this.taskService.getTaskCenterConfigByCode("If_Show_My_Focus", 'all', '*').subscribe(function (re) {
            load.close();
            re && re.currentValue === 'false' && (_this.ifShowFocus = false);
            // 是否关注了该任务(从详情打开)
            if (_this.ifShowFocus && _this.task && _this.showButtons) {
                /** @type {?} */
                var load_1 = _this.loadingService.show();
                _this.taskService.ifFocused(_this.task.processInstanceId).subscribe(function (re) {
                    load_1.close();
                    _this.ifFocused = re && re.length ? true : false;
                }, function () {
                    load_1.close();
                });
            }
        }, function () {
            load.close();
        });
    };
    /**
     * 选中动作按钮
     * @param action
     * @param index
     * @returns
     */
    /**
     * 选中动作按钮
     * @param {?} action
     * @param {?=} index
     * @return {?}
     */
    TaskActionImplComponent.prototype.actionSelect = /**
     * 选中动作按钮
     * @param {?} action
     * @param {?=} index
     * @return {?}
     */
    function (action, index) {
        var _this = this;
        /** @type {?} */
        var subject = new Subject();
        this.selectAction = action;
        if (index) {
            this.index = index;
        }
        /** @type {?} */
        var cmpFactory;
        if (action.typeId === 'wf') {
            cmpFactory = this.cfr.resolveComponentFactory(TaskExecuteComponent);
            this.loadComponentFactory(cmpFactory);
            setTimeout(function () {
                subject.next({ result: true });
            }, 100);
        }
        else {
            System.import(action.impl + '?v=' + new Date().getTime().toString()).then(function (moduleFactories) {
                /** @type {?} */
                var moduleFactory = Object.values(moduleFactories)[0];
                /** @type {?} */
                var moduleRef = moduleFactory.create(_this.injector);
                cmpFactory = moduleRef.instance.getActionComponentFactory(action);
                _this.loadComponentFactory(cmpFactory);
                subject.next({ result: true });
            });
        }
        return subject.asObservable();
    };
    /**
     * 动态加载审批组件
     * @param cmpFactory
     */
    /**
     * 动态加载审批组件
     * @private
     * @param {?} cmpFactory
     * @return {?}
     */
    TaskActionImplComponent.prototype.loadComponentFactory = /**
     * 动态加载审批组件
     * @private
     * @param {?} cmpFactory
     * @return {?}
     */
    function (cmpFactory) {
        /** @type {?} */
        var comment;
        if (this.componentRef && this.componentRef.instance) {
            comment = ((/** @type {?} */ (this.componentRef.instance))).comment;
        }
        this.container.clear();
        this.componentRef = this.container.createComponent(cmpFactory, 0, this.injector);
        this.componentRef.instance.origin = this.showButtons ? ActionOrigin.TaskForm : ActionOrigin.TaskCenter;
        if (this.task && this.componentRef.instance.setTask) {
            this.componentRef.instance.setTask(this.task);
        }
        if (this.componentRef.instance.setAction) {
            this.componentRef.instance.setAction(this.selectAction);
        }
        if (this.componentRef.instance.dialogHeight) {
            this.dialogHeight = this.componentRef.instance.dialogHeight;
        }
        if (this.componentRef.instance.dialogWidth) {
            this.dialogWidth = this.componentRef.instance.dialogWidth;
        }
        // 暂存wf的审批意见
        ((/** @type {?} */ (this.componentRef.instance))).wfCacheComment = comment;
        this.container.insert(this.componentRef.hostView);
    };
    /**
     * 点击提交
     */
    /**
     * 点击提交
     * @return {?}
     */
    TaskActionImplComponent.prototype.submit = /**
     * 点击提交
     * @return {?}
     */
    function () {
        var _this = this;
        //表单前事件
        if (this.showButtons && window['taskCenterBeforeLoad']) {
            this.submitDisable = true;
            /** @type {?} */
            var context = {
                task: this.task,
                action: this.selectAction
            };
            window['taskCenterBeforeLoad'](context).then(function (re) {
                _this.submitDisable = false;
                if (re && re.result == true) {
                    _this.showConfirmDialog();
                }
            }, function (error) {
                _this.submitDisable = false;
                console.error(error);
            });
        }
        else {
            this.showConfirmDialog();
        }
    };
    /**
     * 是否提交前确认
     */
    /**
     * 是否提交前确认
     * @private
     * @return {?}
     */
    TaskActionImplComponent.prototype.showConfirmDialog = /**
     * 是否提交前确认
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.taskService.getTaskCenterConfigByCode('Is_Confirm_Commit').subscribe(function (config) {
            if (config.currentValue === 'true') {
                _this.msgService.question(_this.i18nService("confirm") + _this.selectAction.name + '？', function () {
                    _this.executeSubmit();
                });
            }
            else {
                _this.executeSubmit();
            }
        });
    };
    /**
     * @private
     * @return {?}
     */
    TaskActionImplComponent.prototype.executeSubmit = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.submitDisable = true;
        if (this.selectAction.preEvents && this.selectAction.preEvents.length > 0
            && this.selectAction.preEvents.filter(function (e) { return !e.batch; }).length > 0) {
            /** @type {?} */
            var taskComment_1;
            /** @type {?} */
            var action = (/** @type {?} */ (this.selectAction));
            if (action.typeId === 'wf') {
                // 取意见框里的意见存放在taskComment里
                taskComment_1 = ((/** @type {?} */ (this.componentRef.instance))).comment;
            }
            else {
            }
            this.executeEvents(tslib_1.__spread(this.selectAction.preEvents.filter(function (e) { return !e.batch; })), function () {
                _this.confirmSubmit(taskComment_1);
            }, taskComment_1);
        }
        else {
            this.confirmSubmit();
        }
    };
    /**
     * @private
     * @param {?=} taskComment
     * @return {?}
     */
    TaskActionImplComponent.prototype.confirmSubmit = /**
     * @private
     * @param {?=} taskComment
     * @return {?}
     */
    function (taskComment) {
        var _this = this;
        /** @type {?} */
        var load = this.loadingService.show();
        this.handleSubmit().subscribe(function (re) {
            if (re) {
                load.close();
                if (re.result) {
                    if (_this.selectAction && _this.selectAction.postEvents.length > 0
                        && _this.selectAction.postEvents.filter(function (e) { return !e.batch; }).length > 0) {
                        _this.executeEvents(tslib_1.__spread(_this.selectAction.postEvents.filter(function (e) { return !e.batch; })), function () {
                            _this.submited.emit(re);
                            setTimeout(function () {
                                _this.submitDisable = false;
                            }, 2000);
                        }, taskComment);
                    }
                    else {
                        _this.submited.emit(re);
                        setTimeout(function () {
                            _this.submitDisable = false;
                        }, 2000);
                    }
                }
                else {
                    _this.submitDisable = false;
                }
            }
            _this.actionCodeMap.set('isTransfer', _this.selectAction.code === 'Transfer');
            _this.actionCodeMap.set('isAddSignFront', _this.selectAction.code === 'AddSignFront');
            _this.actionCodeMap.set('isUnClaim', _this.selectAction.code === 'Unclaim');
            _this.actionCodeMap.set('isComment', _this.selectAction.code === 'Comment');
            _this.actionCodeMap.set('isClaim', _this.selectAction.code === 'Claim');
            _this.actionCode.emit(_this.actionCodeMap);
        });
    };
    /**
     * @return {?}
     */
    TaskActionImplComponent.prototype.handleSubmit = /**
     * @return {?}
     */
    function () {
        return this.componentRef.instance.submit();
    };
    /**
     * @param {?} tasks
     * @return {?}
     */
    TaskActionImplComponent.prototype.batchHandleSubmit = /**
     * @param {?} tasks
     * @return {?}
     */
    function (tasks) {
        return this.componentRef.instance.batchSubmit(tasks);
    };
    /**
     * @private
     * @param {?} events
     * @param {?} handler
     * @param {?=} taskComment
     * @return {?}
     */
    TaskActionImplComponent.prototype.executeEvents = /**
     * @private
     * @param {?} events
     * @param {?} handler
     * @param {?=} taskComment
     * @return {?}
     */
    function (events, handler, taskComment) {
        var _this = this;
        if (events && events.length > 0) {
            /** @type {?} */
            var event_1 = events.shift();
            /** @type {?} */
            var url = event_1.eventImpl.split('#')[0];
            /** @type {?} */
            var func_1 = event_1.eventImpl.split('#')[1];
            System.import(url + '?v=' + new Date().getTime().toString())
                .then(function (moduleFactories) {
                /** @type {?} */
                var moduleFactory = Object.values(moduleFactories)[0];
                /** @type {?} */
                var moduleRef = moduleFactory.create(_this.injector);
                /** @type {?} */
                var context = {
                    task: _this.task,
                    action: _this.selectAction,
                    origin: 'TaskForm',
                    formIframe: _this.iframe,
                    comment: taskComment,
                };
                moduleRef.instance[func_1].apply(moduleRef.instance, [context]).subscribe(function (re) {
                    // 兼容旧的布尔类型参数
                    if (typeof re === 'boolean') {
                        if (re) {
                            _this.executeEvents(events, handler, taskComment);
                        }
                        else {
                            _this.submitDisable = false;
                        }
                    }
                    else {
                        if (re.result) {
                            if (re.comment) {
                                taskComment = re.comment;
                                // 取到的前事件意见包含在taskComment里
                            }
                            _this.executeEvents(events, handler, taskComment);
                        }
                        else {
                            _this.submitDisable = false;
                        }
                    }
                });
            });
        }
        else {
            ((/** @type {?} */ (this.componentRef.instance))).comment = taskComment;
            handler();
        }
    };
    /**
     * 点击添加/取消关注
     */
    /**
     * 点击添加/取消关注
     * @return {?}
     */
    TaskActionImplComponent.prototype.focusTask = /**
     * 点击添加/取消关注
     * @return {?}
     */
    function () {
        /** @type {?} */
        var load = this.loadingService.show();
        if (this.ifFocused) {
            this.taskService.addFocus(this.task.sourceId).subscribe(function () {
                load.close();
            }, function () {
                load.close();
            });
        }
        else {
            this.taskService.removeFocus(this.task.processInstanceId).subscribe(function () {
                load.close();
            }, function () {
                load.close();
            });
        }
    };
    /**
     * 国际化
     * @param code
     * @returns
     */
    /**
     * 国际化
     * @param {?} code
     * @return {?}
     */
    TaskActionImplComponent.prototype.i18nService = /**
     * 国际化
     * @param {?} code
     * @return {?}
     */
    function (code) {
        /** @type {?} */
        var result;
        /** @type {?} */
        var supportLanguage = ["zh-CHS", "zh-CHT", "en"];
        /** @type {?} */
        var curLanguage = localStorage.getItem("languageCode");
        /** @type {?} */
        var translate = {
            "zh-CHS": {
                focus: "关注此流程",
                focused: "已关注此流程",
                confirm: "确认"
            },
            "zh-CHT": {
                focus: "關註此流程",
                focused: "已關註此流程",
                confirm: "確認"
            },
            en: {
                focus: "Focus this process",
                focused: "Focused this process",
                confirm: "Confirm"
            },
        };
        if (curLanguage && supportLanguage.includes(curLanguage)) {
            result = translate[curLanguage][code];
        }
        else {
            result = translate["zh-CHS"][code];
        }
        return result;
    };
    TaskActionImplComponent.decorators = [
        { type: Component, args: [{
                    selector: 'task-action-impl',
                    template: "<form\r\n  [class]=\"showButtons ? '' : 'px-4 py-2'\"\r\n  style=\"display: flex; flex-direction: column\"\r\n>\r\n  <div\r\n    *ngIf=\"showButtons\"\r\n    class=\"fa-approval-operate-list fa-approve-common-gutter flex-wrap\"\r\n  >\r\n    <div\r\n      *ngFor=\"let action of actions; let i = index\"\r\n      class=\"fa-approval-operate\"\r\n    >\r\n      <input\r\n        type=\"radio\"\r\n        name=\"operate\"\r\n        [value]=\"i\"\r\n        [id]=\"i\"\r\n        class=\"approval-operate--radio\"\r\n        [checked]=\"i === 0 ? true : false\"\r\n      />\r\n      <label\r\n        class=\"approval-operate--label\"\r\n        [for]=\"i\"\r\n        (click)=\"actionSelect(action, i)\"\r\n      >\r\n        <span class=\"label-title\">{{ action.name }}</span>\r\n        <div class=\"label-tip\"></div>\r\n        <span class=\"f-icon label-icon\"></span>\r\n      </label>\r\n    </div>\r\n  </div>\r\n  <!--\u7EC4\u4EF6\u5BB9\u5668-->\r\n  <ng-container #container></ng-container>\r\n\r\n  <!-- \u5173\u6CE8\u6B64\u6D41\u7A0B -->\r\n  <div\r\n    *ngIf=\"\r\n      showButtons && selectAction && selectAction.typeId === 'wf' && ifShowFocus\r\n    \"\r\n    class=\"custom-control custom-checkbox custom-control-inline\"\r\n    style=\"margin-bottom: 20px\"\r\n  >\r\n    <input\r\n      type=\"checkbox\"\r\n      class=\"custom-control-input\"\r\n      [(ngModel)]=\"ifFocused\"\r\n      (ngModelChange)=\"focusTask()\"\r\n      id=\"taskcenterfocus\"\r\n      name=\"taskcenterfocus\"\r\n    />\r\n    <label class=\"custom-control-label\" for=\"taskcenterfocus\">{{\r\n      focus\r\n    }}</label>\r\n  </div>\r\n\r\n  <!-- \u63D0\u4EA4\u6309\u94AE -->\r\n  <div class=\"fa-btn-group-lg\">\r\n    <button\r\n      *ngIf=\"showButtons && actions.length\"\r\n      type=\"button\"\r\n      class=\"btn btn-primary\"\r\n      [disabled]=\"submitDisable\"\r\n      directive-throttle\r\n      (throttleClick)=\"submit()\"\r\n    >\r\n      {{ submitName }}\r\n    </button>\r\n  </div>\r\n</form>\r\n",
                    styles: [".fa-approval-operate-list{display:flex;flex-direction:row;align-items:center}.fa-approval-operate-list .fa-approval-operate{margin-right:8px;margin-bottom:8px}.fa-approval-operate-list .approval-operate--label{position:relative;padding:4px 21px;overflow:hidden;cursor:pointer;margin-bottom:0;background:#f2f4f7;border-radius:6px;color:#424347;font-weight:500;font-size:13px}.fa-approval-operate-list .approval-operate--label:hover{color:#388fff;border-color:#388fff}.fa-approval-operate-list .approval-operate--label .label-title{line-height:20px;white-space:nowrap}.fa-approval-operate-list .approval-operate--label .label-tip{position:absolute;display:none;right:0;bottom:0;width:14px;height:14px;border:7px solid #dadada;border-top:7px solid transparent;border-left:7px solid transparent}.fa-approval-operate-list .approval-operate--label .label-icon{position:absolute;display:none;right:-3px;bottom:-2px;font-size:12px;color:#fff}.fa-approval-operate-list .approval-operate--label .label-icon::before{content:\"\\e118\"}.fa-approval-operate-list .approval-operate--radio{display:none}.fa-approval-operate-list .approval-operate--radio:checked~.approval-operate--label{padding:3px 19px;font-size:14px;background:rgba(23,145,255,.05);box-shadow:0 4px 10px 0 rgba(69,144,255,.3);border-radius:6px;border:1px solid #388fff;color:#388fff}.fa-approval-operate-list .approval-operate--radio:checked~.approval-operate--label .label-tip{display:block;border-right-color:#388fff;border-bottom-color:#388fff}.fa-approval-operate-list .approval-operate--radio:checked~.approval-operate--label .label-icon{display:block}.fa-btn-like-dropdown{display:flex;align-items:center}.fa-btn-like-dropdown .like-dropdown--text{margin:0 4px 0 0}.fa-btn-like-dropdown .like-dropdown--icon{color:rgba(0,0,0,.25)}.fa-approve-common-gutter{margin-bottom:20px}.fa-approve-comment .approve-comment--title{display:inline-block;color:#667580;border-radius:4px 4px 0 0;height:26px;padding:0 9px;position:relative;background:#f7f7f7;cursor:pointer}.fa-approve-comment .approve-comment--title:after{content:\"\";height:100%;position:absolute;right:-10px;top:0;bottom:0;border-right:5px solid transparent;border-left:5px solid #f7f7f7;border-bottom:13px solid #f7f7f7;border-top:13px solid transparent}.fa-approve-comment .approve-comment--title:hover{background:rgba(23,145,255,.05);color:#1791ff}.fa-approve-comment .approve-comment--title:hover:after{border-bottom-color:rgba(23,145,255,.05);border-left-color:rgba(23,145,255,.05)}.fa-approve-comment .approve-comment--title .fa-btn-like-dropdown{font-size:12px;padding:5px 0}.fa-approve-comment .fa-approve-comment--content .content-textarea{border-color:#dcdcdc;border-radius:0 0 2px 2px;height:140px}.fa-approve-comment .fa-approve-comment--footer{font-size:13px;color:#667580;background:#f7f7f7;padding:10px 0}.fa-approve-comment .approve-comment--text-btns{display:flex;flex-direction:row}.fa-approve-comment .approve-comment--text-btns .text-btns-item{display:flex;flex-direction:row;align-items:center;font-size:12px;line-height:18px;color:#667580;padding:4px;margin-right:8px;cursor:pointer}.fa-approve-comment .approve-comment--text-btns .text-btns-item:hover{color:#388fff}.fa-approve-comment .approve-comment--text-btns .text-btns-item .f-icon{margin-right:6px}.fa-approveal-links::after{content:\".\";display:block;height:0;clear:both;visibility:hidden}.fa-approveal-links .approveal-links--link{font-size:14px;position:relative;padding:0 16px 0 0;display:flex;align-items:center;float:left}.fa-approveal-links .approveal-links--link .link-text{color:rgba(0,0,0,.65);margin:0 8px 0 0}.fa-approveal-links .approveal-links--link+.approveal-links--link{padding-left:16px}.fa-approveal-links .approveal-links--link+.approveal-links--link::before{content:\"\";position:absolute;width:1px;height:80%;left:0;top:10%;background:#dbdbdb}.fa-btn-group-lg .btn{width:180px;height:32px;box-shadow:0 4px 10px 0 rgba(69,144,255,.25);border-radius:6px;min-width:100px;text-align:center;font-size:13px}.fa-approve-record{padding:0 0 16px;position:relative}.fa-approve-record .fa-approve-record--timeline{position:absolute;width:34px;top:0;bottom:0;z-index:100}.fa-approve-record .fa-approve-record--timeline::after{content:\"\";position:absolute;width:1px;height:100%;overflow:hidden;left:50%;top:0;z-index:101;border-left:1px dashed #dcdcdc}.fa-approve-record .fa-approve-record--timeline .fa-approve-star{position:relative;z-index:103;margin:20px auto}.fa-approve-record.fa-state-running .fa-approve-record--timeline::after{top:20px}.fa-approve-record:last-child .fa-approve-record--timeline::after{height:20px}.fa-approve-record .fa-approve-record--content{margin:0 0 0 34px;padding:16px 14px;background:#f9f9f9;border-radius:2px}.fa-approve-record .fa-approve-record--content .approve-record--header{margin:0 0 12px;display:flex;align-items:center}.fa-approve-record .fa-approve-record--content .approve-record--header .header--title{margin-bottom:0;color:rgba(0,0,0,.85)}.fa-approve-record .fa-approve-record--content .approve-record--header>*{margin:0 19px 0 0}.fa-approve-record .approve-record--list-item{display:flex;flex-direction:row}.fa-approve-record .approve-record--list-item+.approve-record--list-item{padding-top:8px}.fa-approve-record .approve-record--list-item .list-item--auxiliary{width:30px;margin:0 10px 0 0}.fa-approve-record .approve-record--list-item .list-item--main{flex-grow:1;flex-shrink:1;flex-basis:auto;padding:6px 0 0;background:#f9f9f9}.fa-approve-record .approve-record--list-item .list-item--content{padding:20px 0 0}.fa-approve-record .approve-record--list-item .list-item--content .list-item--info{font-size:12px;margin-bottom:0}.fa-approve-record .approve-record--list-item .list-item--content .list-item--info .info-label{color:rgba(0,0,0,.65)}.fa-approve-record .approve-record--list-item .list-item--content .list-item--info .info-text{color:rgba(0,0,0,.85)}.fa-approve-record .approve-record--list-item .list-item--header{display:flex;align-items:center}.fa-approve-record .approve-record--list-item .list-item--header>*{margin:0 16px 0 0}.fa-approve-record .approve-record--list-item .list-item--header .list-item--info{margin-bottom:0;font-size:12px}.fa-approve-record .approve-record--list-item .list-item--header .list-item--info .info-label{color:rgba(0,0,0,.35)}.fa-approve-record .approve-record--list-item .list-item--header .list-item--info .info-text{color:rgba(0,0,0,.65)}.fa-approve-record .approve-record--list-item .list-item--name{color:#4297fa}.fa-approve-record .approve-record--avatar{width:30px;height:30px;border-radius:15px}.fa-approve-round-bg{margin:4px auto 0;width:30px;height:30px;background:#59a1ff;border-radius:18px;line-height:30px;text-align:center;white-space:nowrap;font-size:12px;overflow:hidden}.fa-approve-round-bg-text{color:#fff}.fa-approve-star{width:14px;height:14px;display:block;background:rgba(56,143,255,.4);border-radius:7px;position:relative}.fa-approve-star::after{content:\"\";width:8px;height:8px;position:absolute;top:50%;left:50%;margin:-4px 0 0 -4px;border-radius:4px;background:#388fff}.fa-approve-star.bg-warning{background:rgba(245,156,36,.4)!important}.fa-approve-star.bg-warning::after{background:#f59c24}.fa-approve-badge{padding-top:4px;padding-bottom:4px;border-radius:10px}.fa-approve-badge-light{font-size:13px;line-height:20px;padding:0 8px;color:#5f637a;border:1px solid #e6e6e6;background:#f6faff}.fa-approve-link-border{cursor:pointer;display:inline-block;border:1px solid #60a5ff;border-radius:13px;font-size:12px;line-height:16px;color:#388fff;padding:3px 10px;background:#ebf5ff}.fa-approve-link-border:hover{background:#2881ff;color:#fff}"]
                }] }
    ];
    /** @nocollapse */
    TaskActionImplComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: LoadingService },
        { type: MessagerService },
        { type: TaskActionImplService }
    ]; };
    TaskActionImplComponent.propDecorators = {
        container: [{ type: ViewChild, args: ['container', { read: ViewContainerRef },] }],
        actions: [{ type: Input }],
        submitName: [{ type: Input }],
        showButtons: [{ type: Input }],
        batch: [{ type: Input }],
        task: [{ type: Input }],
        iframe: [{ type: Input }],
        origin: [{ type: Input }],
        submited: [{ type: Output }],
        actionCode: [{ type: Output }]
    };
    return TaskActionImplComponent;
}());
export { TaskActionImplComponent };
if (false) {
    /** @type {?} */
    TaskActionImplComponent.prototype.container;
    /** @type {?} */
    TaskActionImplComponent.prototype.actions;
    /** @type {?} */
    TaskActionImplComponent.prototype.submitName;
    /** @type {?} */
    TaskActionImplComponent.prototype.showButtons;
    /** @type {?} */
    TaskActionImplComponent.prototype.batch;
    /** @type {?} */
    TaskActionImplComponent.prototype.task;
    /** @type {?} */
    TaskActionImplComponent.prototype._iframe;
    /** @type {?} */
    TaskActionImplComponent.prototype.origin;
    /** @type {?} */
    TaskActionImplComponent.prototype.submited;
    /** @type {?} */
    TaskActionImplComponent.prototype.actionCode;
    /** @type {?} */
    TaskActionImplComponent.prototype.selectAction;
    /** @type {?} */
    TaskActionImplComponent.prototype.actionCodeMap;
    /** @type {?} */
    TaskActionImplComponent.prototype.componentRef;
    /** @type {?} */
    TaskActionImplComponent.prototype.index;
    /** @type {?} */
    TaskActionImplComponent.prototype.submitDisable;
    /** @type {?} */
    TaskActionImplComponent.prototype.ifFocused;
    /** @type {?} */
    TaskActionImplComponent.prototype.focus;
    /** @type {?} */
    TaskActionImplComponent.prototype.ifShowFocus;
    /** @type {?} */
    TaskActionImplComponent.prototype.dialogHeight;
    /** @type {?} */
    TaskActionImplComponent.prototype.dialogWidth;
    /** @type {?} */
    TaskActionImplComponent.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    TaskActionImplComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    TaskActionImplComponent.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    TaskActionImplComponent.prototype.loadingService;
    /**
     * @type {?}
     * @private
     */
    TaskActionImplComponent.prototype.msgService;
    /**
     * @type {?}
     * @private
     */
    TaskActionImplComponent.prototype.taskService;
}
