// @farris/mobile-devkit/esm2015/lib/ui-state/control-state/control_state_interceptor.service.js
import { shareReplay } from "rxjs/operators";
import { InjectFlags } from "../../core";
import { CONTROL_STATE_INTERCEPTOR_TOKEN, InitContext, ProcessContext } from "./control_state_interceptor";
import { forkJoin, isObservable, of } from "rxjs";
import { EnvUtil, UrlUtil } from "../../utils";
/**
 * 状态拦截器服务
 */
export class ControlStateInterceptorService {
    constructor(injector) {
        this.injector = injector;
        //控件状态拦截器
        this.controlInterceptors = [];
        this.controlInterceptors = this.injector.get(CONTROL_STATE_INTERCEPTOR_TOKEN, null, InjectFlags.Optional);
    }
    init() {
        if (this.result) {
            return this.result;
        }
        const inits = [];
        const initContext = new InitContext();
        const processContext = this.getProcessContext();
        initContext.processContext = this.processContext = processContext;
        for (const controlInterceptor of this.controlInterceptors) {
            if (controlInterceptor.init && typeof controlInterceptor.init === 'function') {
                const initFunc = controlInterceptor.init(initContext);
                const $result = isObservable(initFunc) ? initFunc : of(initFunc);
                inits.push($result);
            }
        }
        return this.result = forkJoin(inits).pipe(shareReplay(1));
    }
    intercept(interceptContext) {
        interceptContext.processContext = this.processContext;
        for (const controlInterceptor of this.controlInterceptors) {
            controlInterceptor.intercept(interceptContext);
        }
    }
    getProcessContext() {
        const processContext = new ProcessContext();
        const isInWf = EnvUtil.isInWf();
        processContext.isInWf = isInWf;
        if (isInWf === false) {
            return processContext;
        }
        const iframe = window.parent.document.querySelector('iframe');
        if (!iframe || !iframe.src) {
            return processContext;
        }
        const queryParams = UrlUtil.getParams(iframe.src);
        if (!queryParams) {
            return processContext;
        }
        processContext.uiStateInProcess = queryParams.UIStateInProcess;
        processContext.formConfigId = queryParams.formConfigId;
        return processContext;
    }
}
