// @farris/mobile-devkit/esm2015/lib/ui-state/control-state/render_state_util.js
/**
 * 零代码状态工具类
 */
class RenderStateUtil {
    /**
     * 制单状态
     * @sumamry
     * 1、发起流程中通过菜单打开；
     * 2、草稿箱列表打开；
     */
    static getRenderStatesWhenDrafting() {
        const renderStates = {
            showHeader: true,
            showChildHeader: true,
            showFooter: true,
            showChildFooter: true,
        };
        return renderStates;
    }
    /**
     * 撤回重新制单
     * @summary
     * 我的发起中打开已撤回并且未重新提交的单据
     */
    static getRenderStatesWhenReDrafting() {
        // 该场景下，重新提交后跳转有问题，暂时不支持
        const renderStates = {
            showHeader: false,
            showChildHeader: false,
            showFooter: true,
            showChildFooter: true,
        };
        // const renderStates = {
        //   showHeader: false,
        //   showChildHeader: false,
        //   showFooter: false,
        //   showChildFooter: false,
        //   canInput: false,
        // };
        return renderStates;
    }
    /**
     * 驳回修改
     * @summary
     * 1、我的代办中共打开已驳回的单据；
     */
    static getRenderStatesWhenBackEdit() {
        const renderStates = {
            showHeader: false,
            showChildHeader: false,
            showFooter: false,
            showChildFooter: true,
        };
        return renderStates;
    }
    /**
     * 审批中
     * @sumamry
     * 1、我的代办中：未被驳回的单据；
     * 2、默认为只读态，最终以流程节点上的配置为准
     */
    static getRenderStatesWhenApproving() {
        const renderStates = {
            showHeader: false,
            showChildHeader: false,
            showFooter: false,
            showChildFooter: false,
        };
        return renderStates;
    }
    /**
     * 仅查看
     */
    static getRenderStatesWhenViewOnly() {
        const renderStates = {
            showHeader: false,
            showFooter: false,
            showChildHeader: false,
            showChildFooter: false,
        };
        return renderStates;
    }
}
export { RenderStateUtil as NoCodeRenderStateUtil };
