// @farris/ui-tabs/esm5/lib/tabs.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { FDropdownDirectiveTypeModule } from '@farris/ui-dropdown';
import { NgbDropdownModule } from './dropdown/dropdown.module';
import { NgbTooltipModule } from './tooltip/tooltip.module';
import { LocaleModule } from '@farris/ui-locale';
import { FarrisTabsComponent } from './tabs/tabs.component';
import { TabConfig, TabService } from './tabs.config';
import { FarrisTabComponent } from './tab/tab.component';
import { TabTranscludeDirective } from './tab/tab.transclude.directive';
import { TabHeadDirective } from './tab/tab.headtemplate.directive';
import { TabHostDirective } from './tab/tabhost.directive';
import { FarrisTabsExtendDirective } from './tab/extend.directive';
var FarrisTabsModule = /** @class */ (function () {
    function FarrisTabsModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    FarrisTabsModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: FarrisTabsModule,
            providers: [{ provide: TabConfig, useValue: config }]
        };
    };
    FarrisTabsModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        LocaleModule,
                        FormsModule,
                        NgbDropdownModule.forRoot(),
                        NgbTooltipModule.forRoot(),
                        FDropdownDirectiveTypeModule
                    ],
                    declarations: [FarrisTabsComponent, FarrisTabComponent,
                        TabTranscludeDirective, TabHeadDirective, FarrisTabsExtendDirective, TabHostDirective],
                    exports: [
                        FarrisTabsComponent,
                        FarrisTabComponent,
                        TabTranscludeDirective,
                        TabHeadDirective,
                        TabHostDirective,
                        FarrisTabsExtendDirective,
                        NgbDropdownModule,
                        NgbTooltipModule
                    ],
                    providers: [TabService]
                },] }
    ];
    return FarrisTabsModule;
}());
export { FarrisTabsModule };
