// @farris/ui-tabs/esm5/lib/tabs.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
var TabConfig = /** @class */ (function () {
    function TabConfig() {
        // 是否启用下拉框
        this.showDropDown = true;
        // 是否启用提示信息
        this.showTooltips = true;
        // 点击时滚动值
        this.scrollStep = 15;
        // 是否自动重新宽度 ---- todo:此处只处理了window的resize事件
        this.autoResize = true;
        // 下拉列表最大高度 非正值表明是不高度
        this.maxDropdownH = 200;
        // 下拉列表最大宽度： 非正值表明是不限制宽度
        this.maxDropdownW = -1;
        // 标签最大宽度：非正值表明不限制宽度
        this.maxTabW = 100;
        // 提示信息条数
        this.maxTooltipLen = 5;
    }
    return TabConfig;
}());
export { TabConfig };
if (false) {
    /** @type {?} */
    TabConfig.prototype.showDropDown;
    /** @type {?} */
    TabConfig.prototype.showTooltips;
    /** @type {?} */
    TabConfig.prototype.scrollStep;
    /** @type {?} */
    TabConfig.prototype.autoResize;
    /** @type {?} */
    TabConfig.prototype.maxDropdownH;
    /** @type {?} */
    TabConfig.prototype.maxDropdownW;
    /** @type {?} */
    TabConfig.prototype.maxTabW;
    /** @type {?} */
    TabConfig.prototype.maxTooltipLen;
}
var TabService = /** @class */ (function () {
    function TabService(exConfig) {
        this.config = new TabConfig();
        if (exConfig) {
            Object.assign(this.config, exConfig);
        }
    }
    /**
     * @return {?}
     */
    TabService.prototype.getConfig = /**
     * @return {?}
     */
    function () {
        return this.config;
    };
    TabService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    TabService.ctorParameters = function () { return [
        { type: TabConfig, decorators: [{ type: Optional }] }
    ]; };
    return TabService;
}());
export { TabService };
if (false) {
    /** @type {?} */
    TabService.prototype.config;
}
