// @farris/ide-devkit/esm5/lib/ideframework/package/package.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PackageView } from './package-view';
var Package = /** @class */ (function () {
    function Package(gsp) {
        this.gsp = gsp;
        // leftPanel: any;
        // modelPanel: any;
        this.subscriptions = [];
        this.packageId = gsp.util.newGuid();
    }
    Object.defineProperty(Package.prototype, "id", {
        get: /**
         * @return {?}
         */
        function () {
            return this.packageId;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.packageId = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Package.prototype, "view", {
        get: /**
         * @return {?}
         */
        function () {
            return this.packageView;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} state
     * @return {?}
     */
    Package.prototype.activate = /**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this.packageView = new PackageView(state.sagipackageViewState);
        // /*this.leftPanel = */this.ws.addLeftPanel({
        //   title: 'test',
        //   item: this.packageView.getComponent(),
        //   visible: false
        // });
    };
    /**
     * @return {?}
     */
    Package.prototype.initialize = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    Package.prototype.deactivate = /**
     * @return {?}
     */
    function () {
        this.packageView.destroy();
    };
    /**
     * @return {?}
     */
    Package.prototype.serialize = /**
     * @return {?}
     */
    function () {
        return {
            sagipackageViewState: this.packageView.serialize()
        };
    };
    /**
     * @return {?}
     */
    Package.prototype.toggle = /**
     * @return {?}
     */
    function () {
        // return (
        //   this.modelPanel.isVisible() ?
        //     this.modelPanel.hide() :
        //     this.modelPanel.show()
        // );
    };
    /**
     * @return {?}
     */
    Package.prototype.getComponent = /**
     * @return {?}
     */
    function () {
        return null;
    };
    return Package;
}());
export { Package };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    Package.prototype.packageView;
    /**
     * @type {?}
     * @protected
     */
    Package.prototype.subscriptions;
    /**
     * @type {?}
     * @protected
     */
    Package.prototype.packageId;
    /** @type {?} */
    Package.prototype.data;
    /**
     * @type {?}
     * @protected
     */
    Package.prototype.gsp;
}
