// @edp-aif/common-api/esm2015/lib/content/impl/BizEventContent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Content } from './Content';
import { BizEventEntity } from '../../entity/impl/BizEventEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
export class BizEventContent extends Content {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.canDoExpression = null;
        this.bizFlowchartIdExpression = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "canDoExpression", this.canDoExpression);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartIdExpression", this.bizFlowchartIdExpression);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.canDoExpression = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "canDoExpression");
        this.bizFlowchartIdExpression = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartIdExpression");
    }
    /**
     * @protected
     * @return {?}
     */
    GetEntityClass() {
        return BizEventEntity;
    }
}
if (false) {
    /** @type {?} */
    BizEventContent.prototype.canDoExpression;
    /** @type {?} */
    BizEventContent.prototype.bizFlowchartIdExpression;
}
