// @edp-aif/common-api/esm2015/lib/content/impl/BizFlowchartContent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Content } from './Content';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
import { BizFlowchartEntity } from '../../entity/impl/BizFlowchartEntity';
export class BizFlowchartContent extends Content {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.bpmnModel = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "bpmnModel", this.bpmnModel);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.bpmnModel = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bpmnModel");
    }
    /**
     * @protected
     * @return {?}
     */
    GetEntityClass() {
        return BizFlowchartEntity;
    }
}
if (false) {
    /** @type {?} */
    BizFlowchartContent.prototype.bpmnModel;
}
