// @edp-aif/common-api/esm2015/lib/content/impl/Content.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl } from '../../JsonConvertableImpl';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * @abstract
 * @template TEntity
 */
export class Content extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.entity = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "entity", this.entity);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.entity = BpmnModelJsonHelper.GetAttributeValue(obj, "entity");
    }
}
if (false) {
    /** @type {?} */
    Content.prototype.entity;
    /**
     * @abstract
     * @protected
     * @template ClassOfT
     * @return {?}
     */
    Content.prototype.GetEntityClass = function () { };
}
