// @edp-aif/common-api/esm2015/lib/content/impl/CreationRuleContent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Content } from './Content';
import { CreationRuleEntity } from '../../entity/impl/CreationRuleEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
export class CreationRuleContent extends Content {
    /**
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.mappingContent = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingContent", this.mappingContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "formSettings", this.formSettings);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.mappingContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingContent");
        this.formSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettings");
    }
    /**
     * @protected
     * @return {?}
     */
    GetEntityClass() {
        return CreationRuleEntity;
    }
}
if (false) {
    /** @type {?} */
    CreationRuleContent.prototype.mappingContent;
    /** @type {?} */
    CreationRuleContent.prototype.formSettings;
}
