// @edp-aif/common-api/esm2015/lib/content/impl/WritebackRuleContent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Content } from './Content';
import { WritebackRuleEntity } from '../../entity/impl/WritebackRuleEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
export class WritebackRuleContent extends Content {
    /**
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.mappingContent = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingContent", this.mappingContent);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.mappingContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingContent");
    }
    /**
     * @protected
     * @return {?}
     */
    GetEntityClass() {
        return WritebackRuleEntity;
    }
}
if (false) {
    /** @type {?} */
    WritebackRuleContent.prototype.mappingContent;
}
