// @gspwidget/util/esm5/lib/types/common.type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 支持的语言
 */
var /**
 * 支持的语言
 */
Lang = /** @class */ (function () {
    function Lang(code, name) {
        this.code = code;
        this.name = name;
    }
    return Lang;
}());
/**
 * 支持的语言
 */
export { Lang };
if (false) {
    /** @type {?} */
    Lang.prototype.code;
    /** @type {?} */
    Lang.prototype.name;
}
var I18nValue = /** @class */ (function () {
    function I18nValue() {
    }
    return I18nValue;
}());
export { I18nValue };
/**
 * @record
 */
export function ErrorInfo() { }
if (false) {
    /** @type {?} */
    ErrorInfo.prototype.timestamp;
    /** @type {?} */
    ErrorInfo.prototype.status;
    /** @type {?} */
    ErrorInfo.prototype.error;
    /** @type {?} */
    ErrorInfo.prototype.message;
    /** @type {?} */
    ErrorInfo.prototype.path;
    /** @type {?} */
    ErrorInfo.prototype.trace;
}
/**
 * @record
 */
export function TreeNode() { }
if (false) {
    /** @type {?|undefined} */
    TreeNode.prototype.children;
    /** @type {?} */
    TreeNode.prototype.title;
    /** @type {?} */
    TreeNode.prototype.key;
    /** @type {?|undefined} */
    TreeNode.prototype.value;
    /** @type {?|undefined} */
    TreeNode.prototype.label;
    /** @type {?|undefined} */
    TreeNode.prototype.isLeaf;
    /* Skipping unhandled member: [key: string]: any*/
}
