// @gspwidget/util/esm5/lib/types/page.type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 页面的保存结构
 */
var /**
 * 页面的保存结构
 */
PageInfo = /** @class */ (function () {
    // TODO: 希望“页面管理”里，服务端返回的列表中的 page 加上当前页面分配给的功能组列表，如果性能没问题的话
    function PageInfo(id) {
        this.id = id;
        this.name = '';
        this.i18nName = {};
        this.config = new PageConfig();
    }
    return PageInfo;
}());
/**
 * 页面的保存结构
 */
export { PageInfo };
if (false) {
    /** @type {?} */
    PageInfo.prototype.name;
    /** @type {?} */
    PageInfo.prototype.i18nName;
    /** @type {?} */
    PageInfo.prototype.config;
    /** @type {?} */
    PageInfo.prototype.creator;
    /** @type {?} */
    PageInfo.prototype.createTime;
    /** @type {?} */
    PageInfo.prototype.modifier;
    /** @type {?} */
    PageInfo.prototype.modifyTime;
    /** @type {?} */
    PageInfo.prototype.pageType;
    /** @type {?} */
    PageInfo.prototype.id;
}
/**
 * 运行时页面得到的页面列表，是这么个简单玩意
 * @record
 */
export function SimplePage() { }
if (false) {
    /** @type {?} */
    SimplePage.prototype.pageId;
    /** @type {?} */
    SimplePage.prototype.pageName;
    /** @type {?} */
    SimplePage.prototype.selected;
    /** @type {?|undefined} */
    SimplePage.prototype.pageType;
}
/**
 * 拖到页面上的部件实例，前端生成 id
 * @record
 */
export function PortletInstance() { }
if (false) {
    /** @type {?} */
    PortletInstance.prototype.instanceId;
    /** @type {?} */
    PortletInstance.prototype.isBAPortlet;
    /** @type {?} */
    PortletInstance.prototype.gridsterItem;
    /** @type {?} */
    PortletInstance.prototype.portletId;
    /**
     * 联动配置
     * @type {?|undefined}
     */
    PortletInstance.prototype.link;
    /**
     * 来自前端：
     * 1. 设计时拖入
     * 2. 加载完成赋值
     * 真正保存时不存它。
     * @type {?|undefined}
     */
    PortletInstance.prototype.portlet;
}
/**
 * @record
 */
export function LinkMap() { }
var PageConfig = /** @class */ (function () {
    function PageConfig() {
        this.type = 'gridster';
        /**
         * 当 type 为 gridster，布局信息
         */
        this.layout = [];
        /** TODO: 当前页面其他配置，比如 gridster options */
    }
    return PageConfig;
}());
export { PageConfig };
if (false) {
    /** @type {?} */
    PageConfig.prototype.type;
    /**
     * 当 type 为 gridster，布局信息
     * @type {?}
     */
    PageConfig.prototype.layout;
    /**
     * 当 type 为 func，菜单 id
     * @type {?}
     */
    PageConfig.prototype.funcId;
}
/**
 * @record
 */
export function PageWithRole() { }
if (false) {
    /** @type {?} */
    PageWithRole.prototype.id;
    /** @type {?} */
    PageWithRole.prototype.roleId;
    /** @type {?} */
    PageWithRole.prototype.roleName;
    /** @type {?} */
    PageWithRole.prototype.pageId;
    /** @type {?} */
    PageWithRole.prototype.pageName;
    /** @type {?} */
    PageWithRole.prototype.createTime;
    /** @type {?} */
    PageWithRole.prototype.creator;
    /** @type {?} */
    PageWithRole.prototype.enable;
}
