// @gspwidget/util/esm5/lib/types/portlet.type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/** @type {?} */
export var DEFAULT_PORTLET_COLS = 5;
/** @type {?} */
export var DEFAULT_PORTLET_ROWS = 4;
/**
 * 部件
 * @abstract
 */
var /**
 * 部件
 * @abstract
 */
Portlet = /** @class */ (function () {
    function Portlet() {
    }
    /**
     * 唯一标识
     * @param {?} key
     * @return {?}
     */
    Portlet.prototype.getSize = /**
     * 唯一标识
     * @param {?} key
     * @return {?}
     */
    function (key) {
        if (key === 'rows') {
            return this._getAxisSize('rows') || this._getAxisSize('minRows') || DEFAULT_PORTLET_ROWS;
        }
        else if (key === 'cols') {
            return this._getAxisSize('cols') || this._getAxisSize('minCols') || DEFAULT_PORTLET_COLS;
        }
        else {
            return this._getAxisSize(key);
        }
    };
    return Portlet;
}());
/**
 * 部件
 * @abstract
 */
export { Portlet };
if (false) {
    /** @type {?} */
    Portlet.prototype.id;
    /**
     * @abstract
     * @param {?} key
     * @return {?}
     */
    Portlet.prototype._getAxisSize = function (key) { };
}
var EppPortlet = /** @class */ (function (_super) {
    tslib_1.__extends(EppPortlet, _super);
    function EppPortlet(widgetId) {
        var _this = _super.call(this) || this;
        _this.widgetId = widgetId;
        _this.name = '未命名部件'; /** 显示标题 */
        /**
         * 显示标题
         */
        _this.i18nName = {};
        _this.propValues = {}; /** 可配置属性的值 */
        /**
         * 可配置属性的值
         */
        _this.dataOption = {};
        /** 所属模块 */
        /**
         * 配置联动时，前端动态增加。只做运行时使用，不应该被保存。
         * null 代表初始状态，[] 代表数据已加载且为空。
         */
        _this.dps = null;
        return _this;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    EppPortlet.prototype._getAxisSize = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return (this.size && this.size[key]) || (this.widget.layout && this.widget.layout.grid && this.widget.layout.grid[key]);
    };
    return EppPortlet;
}(Portlet));
export { EppPortlet };
if (false) {
    /** @type {?} */
    EppPortlet.prototype.name;
    /**
     * 显示标题
     * @type {?}
     */
    EppPortlet.prototype.i18nName;
    /** @type {?} */
    EppPortlet.prototype.desc;
    /**
     * 备注
     * @type {?}
     */
    EppPortlet.prototype.size;
    /**
     * 尺寸范围
     * @type {?}
     */
    EppPortlet.prototype.widget;
    /** @type {?} */
    EppPortlet.prototype.propValues;
    /**
     * 可配置属性的值
     * @type {?}
     */
    EppPortlet.prototype.dataOption;
    /** @type {?} */
    EppPortlet.prototype.moduleId;
    /**
     * 配置联动时，前端动态增加。只做运行时使用，不应该被保存。
     * null 代表初始状态，[] 代表数据已加载且为空。
     * @type {?}
     */
    EppPortlet.prototype.dps;
    /**
     * 元部件 Id
     * @type {?}
     */
    EppPortlet.prototype.widgetId;
}
/** @type {?} */
var defaultSizeMap = {
    rows: 4,
    cols: 5,
    minCols: 2,
    minRows: 2,
    maxCols: 24,
    maxRows: 24,
};
var BAPortlet = /** @class */ (function (_super) {
    tslib_1.__extends(BAPortlet, _super);
    function BAPortlet() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isBA = true;
        return _this;
    }
    // code?: string
    // createTime?: string
    // detail?: number
    // gradeCode?: string
    // isParent?: boolean
    // layer?: number
    // pid?: string
    // tab?: number
    // type?: string
    // updateTime?: string
    // userId?: string
    // code?: string
    // createTime?: string
    // detail?: number
    // gradeCode?: string
    // isParent?: boolean
    // layer?: number
    // pid?: string
    // tab?: number
    // type?: string
    // updateTime?: string
    // userId?: string
    /**
     * @param {?} key
     * @return {?}
     */
    BAPortlet.prototype._getAxisSize = 
    // code?: string
    // createTime?: string
    // detail?: number
    // gradeCode?: string
    // isParent?: boolean
    // layer?: number
    // pid?: string
    // tab?: number
    // type?: string
    // updateTime?: string
    // userId?: string
    /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return defaultSizeMap[key];
    };
    return BAPortlet;
}(Portlet));
export { BAPortlet };
if (false) {
    /** @type {?} */
    BAPortlet.prototype.isBA;
    /** @type {?} */
    BAPortlet.prototype.name;
}
