// @gsp-lcm/metadatart-selector/esm5/metadata/metadata.rtservice.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata.rtservice.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { Inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
var GSPMetadataRTService = /** @class */ (function () {
    function GSPMetadataRTService(httpClient, url) {
        this.httpClient = httpClient;
        this.url = url;
        // tslint:disable-next-line:variable-name
        this.MetadataRT_Base_Path = '/api/runtime/lcm/v1.0/rt-metadatas';
        // tslint:disable-next-line:variable-name
        this.MetadataBizType_Base_Path = '/api/runtime/lcm/v1.0/metadata-biztypes';
        this.Server_IP = (url === null || url === undefined) ? 'http://10.24.19.140:5000' : url;
    }
    /**
     * 元数据运行时服务
     */
    /**
     * 元数据运行时服务
     * @param {?} metadataID
     * @return {?}
     */
    GSPMetadataRTService.prototype.GetMetadataRT = /**
     * 元数据运行时服务
     * @param {?} metadataID
     * @return {?}
     */
    function (metadataID) {
        /** @type {?} */
        var url = this.Server_IP + this.MetadataRT_Base_Path + '/' + metadataID;
        return this.httpClient.get(url);
    };
    /**
     * @param {?} typeList
     * @return {?}
     */
    GSPMetadataRTService.prototype.GetMetadataRTList = /**
     * @param {?} typeList
     * @return {?}
     */
    function (typeList) {
        /** @type {?} */
        var url = this.Server_IP + this.MetadataRT_Base_Path + '?metadataTypes=' + typeList;
        return this.httpClient.get(url);
    };
    /**
     * @param {?} typeList
     * @return {?}
     */
    GSPMetadataRTService.prototype.GetMetadataRTListForHelp = /**
     * @param {?} typeList
     * @return {?}
     */
    function (typeList) {
        /** @type {?} */
        var url = this.Server_IP + this.MetadataRT_Base_Path + '/helper?metadataTypes=' + typeList;
        return this.httpClient.get(url);
    };
    /**
     * @param {?} bizTypeId
     * @param {?} metadataTypes
     * @return {?}
     */
    GSPMetadataRTService.prototype.getMetadataListByBizTypeId = /**
     * @param {?} bizTypeId
     * @param {?} metadataTypes
     * @return {?}
     */
    function (bizTypeId, metadataTypes) {
        /** @type {?} */
        var url = this.Server_IP + this.MetadataBizType_Base_Path + '?bizTypeId=' + bizTypeId + '&metadataTypes=' + metadataTypes;
        return this.httpClient.get(url);
    };
    /**
     * @param {?} boList
     * @param {?} bizTypeId
     * @param {?} metadataTypes
     * @return {?}
     */
    GSPMetadataRTService.prototype.getMetadataListByBizTypeIdForHelp = /**
     * @param {?} boList
     * @param {?} bizTypeId
     * @param {?} metadataTypes
     * @return {?}
     */
    function (boList, bizTypeId, metadataTypes) {
        /** @type {?} */
        var url = this.Server_IP + this.MetadataBizType_Base_Path + '/helper?boList=' + boList + '&bizTypeId=' + bizTypeId + '&metadataTypes=' + metadataTypes;
        return this.httpClient.get(url);
    };
    // 获取元数据类型及后缀信息
    // 获取元数据类型及后缀信息
    /**
     * @return {?}
     */
    GSPMetadataRTService.prototype.GetMetadataTypeList = 
    // 获取元数据类型及后缀信息
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var url = this.Server_IP + this.MetadataRT_Base_Path + '/typeList';
        return this.httpClient.get(url);
    };
    // 获取后缀
    // 获取后缀
    /**
     * @param {?} metadataType
     * @return {?}
     */
    GSPMetadataRTService.prototype.GetPostfix = 
    // 获取后缀
    /**
     * @param {?} metadataType
     * @return {?}
     */
    function (metadataType) {
        this.GetMetadataTypeList().subscribe((/**
         * @param {?} typeList
         * @return {?}
         */
        function (typeList) {
            typeList.forEach((/**
             * @param {?} type
             * @return {?}
             */
            function (type) {
                if (metadataType.toLowerCase() === type.typeCode.toLowerCase()) {
                    return type.postfix;
                }
            }));
        }));
        return new Observable();
    };
    GSPMetadataRTService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    GSPMetadataRTService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: String, decorators: [{ type: Inject, args: ['url',] }] }
    ]; };
    return GSPMetadataRTService;
}());
export { GSPMetadataRTService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    GSPMetadataRTService.prototype.Server_IP;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataRTService.prototype.MetadataRT_Base_Path;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataRTService.prototype.MetadataBizType_Base_Path;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataRTService.prototype.httpClient;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataRTService.prototype.url;
}
