// @gsp-lcm/metadatart-selector/esm5/metadata/metadata-rtservice.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-rtservice.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { GSPMetadataRTService } from './metadata.rtservice';
import { HttpClient, HttpClientModule } from '@angular/common/http';
/**
 * @param {?} http
 * @param {?} url
 * @return {?}
 */
export function factory(http, url) {
    return new GSPMetadataRTService(http, url);
}
var GSPMetadataRTServiceModule = /** @class */ (function () {
    function GSPMetadataRTServiceModule() {
    }
    /**
     * @param {?} url
     * @return {?}
     */
    GSPMetadataRTServiceModule.forRoot = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return {
            ngModule: GSPMetadataRTServiceModule,
            providers: [
                { provide: 'url', useValue: url },
                { provide: GSPMetadataRTService, useFactory: factory, deps: [HttpClient, 'url'] }
            ]
        };
    };
    GSPMetadataRTServiceModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        HttpClientModule
                    ],
                    exports: [],
                    providers: []
                },] }
    ];
    return GSPMetadataRTServiceModule;
}());
export { GSPMetadataRTServiceModule };
