// @gsp-wf/wf-process-editor/esm2015/lib/views/vo-config/vo-config.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ComponentFactoryResolver, EventEmitter, Inject, Injector, Input, LOCALE_ID, Optional, Output, TemplateRef, ViewChild } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { LoadingService } from '@farris/ui-loading';
import { MessagerService } from '@farris/ui-messager';
import { BsModalService } from '@farris/ui-modal';
import { NotifyService } from '@farris/ui-notify';
import { SelectMetadataRTComponent } from '@gsp-lcm/metadatart-selector';
import { FrameworkService } from '@gsp-sys/rtf-common';
import { BillTypeHelpComponent } from '../bill-type-help/bill-type-help/bill-type-help.component';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../i18n/pipe/wf-process-editor-locale.pipe';
export class VoConfigComponent {
    /**
     * @param {?} messager
     * @param {?} notify
     * @param {?} modal
     * @param {?} framework
     * @param {?} router
     * @param {?} activatedRoute
     * @param {?} loading
     * @param {?} cfr
     * @param {?} inj
     * @param {?} cdr
     * @param {?} localeId
     * @param {?} resources
     * @param {?} localePipe
     */
    constructor(messager, notify, modal, framework, router, activatedRoute, loading, cfr, inj, cdr, localeId, resources, localePipe) {
        this.messager = messager;
        this.notify = notify;
        this.modal = modal;
        this.framework = framework;
        this.router = router;
        this.activatedRoute = activatedRoute;
        this.loading = loading;
        this.cfr = cfr;
        this.inj = inj;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.groupIcon = '<i class="f-icon f-icon-lookup"></i>';
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.editorParams = {};
        this.modalConfig = {
            title: this.localePipe.transform("component.voConfig.documentInfo"),
            width: 500,
            height: 240,
            showButtons: true
        };
        this.entity = {};
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
    }
    /**
     * @param {?} data
     * @return {?}
     */
    set value(data) {
        if (data) {
            this.entity = data || {};
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    clickOk() {
        /** @type {?} */
        const obj = { value: this.entity, parameters: null };
        this.submitModal.emit(obj);
    }
    /**
     * @return {?}
     */
    clickCancel() {
        this.closeModal.emit();
    }
    /**
     * @return {?}
     */
    selectVo() {
        /** @type {?} */
        const compFactory = this.cfr.resolveComponentFactory(SelectMetadataRTComponent);
        /** @type {?} */
        const compRef = compFactory.create(this.inj);
        compRef.instance.typeList = 'GSPViewModel';
        // 元数据选择器bizTypeId表示业务种类Id
        compRef.instance.bizTypeId = this.editorParams && this.editorParams.bizTypeId;
        compRef.instance.selectedMetadata.subscribe((/**
         * @param {?} metadata4RefDto
         * @return {?}
         */
        (metadata4RefDto) => {
            this.clearVo();
            /** @type {?} */
            const voDto = metadata4RefDto.metadata;
            /** @type {?} */
            const vo = JSON.parse(voDto.content);
            this.entity.currentVoId = vo.ID;
            this.entity.currentVoName = vo.Name;
        }), (/**
         * @param {?} err
         * @return {?}
         */
        (err) => {
            console.error(err);
            // todo 确定异常结构, 目前采用兼容取值方式
            this.messager.error(err.error == null ? err.message : (err.error.Message || err.error.message || err.message));
            this.clearVo();
        }));
        /** @type {?} */
        const modalConfig = {
            title: this.localePipe.transform("component.voConfig.viewModelSelection"),
            width: 930,
            height: 573,
            showButtons: false
        };
        /** @type {?} */
        const dialog = this.modal.show(compRef, modalConfig);
        compRef.instance.closeModalEmitter.subscribe((/**
         * @param {?} $event
         * @return {?}
         */
        $event => dialog.close()));
    }
    /**
     * @return {?}
     */
    clearVo() {
        this.entity.currentVoId = undefined;
        this.entity.currentVoName = undefined;
    }
    /**
     * @return {?}
     */
    selectDocuType() {
        /** @type {?} */
        const cmpF = this.cfr.resolveComponentFactory(BillTypeHelpComponent);
        /** @type {?} */
        const cmpR = cmpF.create(this.inj, []);
        cmpR.instance.docuTypeId = this.entity.currentDocumentTypeId;
        /** @type {?} */
        const buttons = [];
        buttons.push({
            text: this.localePipe.transform("component.voConfig.cancel"),
            cls: "btn btn-secondary",
            handle: (/**
             * @return {?}
             */
            () => {
                bsModal && bsModal.close();
            })
        });
        buttons.push({
            text: this.localePipe.transform("component.voConfig.clickOk"),
            cls: "btn btn-primary",
            handle: (/**
             * @return {?}
             */
            () => {
                /** @type {?} */
                const docuType = cmpR.instance.getSelectedDocuType();
                if (docuType == null) {
                    this.notify.warning(this.localePipe.transform("component.voConfig.selectOne"));
                    return;
                }
                this.entity.currentDocumentTypeId = docuType.id;
                this.entity.currentDocumentTypeName = docuType.name;
                bsModal && bsModal.close();
            })
        });
        /** @type {?} */
        const bsModal = this.modal.show(cmpR, {
            title: this.localePipe.transform("component.voConfig.documentTypeSelect"),
            height: 573,
            width: 930,
            buttons: buttons,
            showMaxButton: false,
        });
    }
    /**
     * @return {?}
     */
    clearDocuType() {
        this.entity.currentDocumentTypeId = undefined;
        this.entity.currentDocumentTypeName = undefined;
    }
}
VoConfigComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-vo-config',
                template: "<div class=\"f-page f-page-card f-page-is-mainsubcard\" id=\"bif-manager-list\">\r\n  <div class=\"f-page-main\" id=\"page-main\" style=\"overflow-y: auto;\">\r\n    <div class=\"f-form-layout farris-form farris-form-controls-inline\">\r\n      <div class=\"col-12\">\r\n        <div class=\"farris-group-wrap\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\" title=\"{{'component.voConfig.viewModel' | wfLocale}}\" for=\"vo\">\r\n              <span class=\"farris-label-text\">{{'component.voConfig.viewModel' | wfLocale}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input-group id=\"vo\" [ngModel]=\"entity&&entity.currentVoName\" [editable]=\"false\" [enableClear]=\"false\"\r\n                (clickHandle)=\"selectVo()\" (clear)=\"clearVo()\" [groupText]=\"groupIcon\" placeholder=\"{{'component.voConfig.placeholder' | wfLocale}}\">\r\n              </input-group>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n\r\n      <div class=\"col-12\">\r\n        <div class=\"farris-group-wrap\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\" title=\"{{'component.voConfig.documenttype' | wfLocale}}\" for=\"vo\">\r\n              <span class=\"farris-label-text\">{{'component.voConfig.documenttype' | wfLocale}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input-group id=\"vo\" [ngModel]=\"entity&&entity.currentDocumentTypeName\" [editable]=\"false\"\r\n                [enableClear]=\"false\" (clickHandle)=\"selectDocuType()\" (clear)=\"clearDocuType()\" [groupText]=\"groupIcon\"\r\n                placeholder=\"{{'component.voConfig.placeholder' | wfLocale}}\">\r\n              </input-group>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm px-3 mx-2\" (click)=\"clickCancel()\">{{'component.voConfig.cancel' | wfLocale}}</button>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm px-3 mx-2\" (click)=\"clickOk()\">{{'component.voConfig.clickOk' | wfLocale}}</button>\r\n</ng-template>\r\n",
                providers: [
                    WfLocalePipe,
                ]
            }] }
];
/** @nocollapse */
VoConfigComponent.ctorParameters = () => [
    { type: MessagerService },
    { type: NotifyService },
    { type: BsModalService },
    { type: FrameworkService },
    { type: Router },
    { type: ActivatedRoute },
    { type: LoadingService },
    { type: ComponentFactoryResolver },
    { type: Injector },
    { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
    { type: WfLocalePipe, decorators: [{ type: Optional }] }
];
VoConfigComponent.propDecorators = {
    closeModal: [{ type: Output }],
    submitModal: [{ type: Output }],
    value: [{ type: Input }],
    editorParams: [{ type: Input }],
    modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }]
};
if (false) {
    /** @type {?} */
    VoConfigComponent.prototype.groupIcon;
    /** @type {?} */
    VoConfigComponent.prototype.closeModal;
    /** @type {?} */
    VoConfigComponent.prototype.submitModal;
    /** @type {?} */
    VoConfigComponent.prototype.editorParams;
    /** @type {?} */
    VoConfigComponent.prototype.modalFooter;
    /** @type {?} */
    VoConfigComponent.prototype.modalConfig;
    /** @type {?} */
    VoConfigComponent.prototype.entity;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.messager;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.modal;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.framework;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.loading;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.inj;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    VoConfigComponent.prototype.localePipe;
}
export class VoConfigConverter {
    /**
     * @param {?} localePipe
     */
    constructor(localePipe) {
        this.localePipe = localePipe;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    convertTo(data) {
        if (data && data.currentVoId) {
            return data.currentVoName || data.currentVoId || this.localePipe.transform("component.voConfig.set");
        }
        return this.localePipe.transform("component.voConfig.notSet");
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    VoConfigConverter.prototype.localePipe;
}
