// @gsp-cmp/ubcc-param-data-ui/esm2015/lib/controls/param-control/date-control/date-control.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/controls/param-control/date-control/date-control.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
export class DateControlComponent {
    constructor() {
        this.canEdit = true;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.param = this.param || {};
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    dateTimeValueChange($event) {
        this.param.value = $event.jsDate;
    }
    /**
     * @return {?}
     */
    isReadonly() {
        if (!this.canEdit) {
            return true;
        }
        else {
            return !this.param || !this.param.readonly ? false : true;
        }
    }
    /**
     * @return {?}
     */
    show() {
        return !this.param || !this.param.description ? false : true;
    }
}
DateControlComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-date-control',
                template: "<div class=\"farris-form-controls-inline farris-group-wrap col-12\">\n  <div class=\"form-group farris-form-group\">\n    <label class=\"col-form-label col-4\">\n      <span class=\"farris-label-text\" style=\"float:right\">{{param.name}}</span>\n    </label>\n    <div class=\"farris-input-wrap col-6\">\n      <farris-datepicker [editable]=\"false\" [readonly]=\"isReadonly()\" [(ngModel)]=\"param.value\" name=\"{{param.code}}\" (valueChange)=\"dateTimeValueChange($event)\">\n      </farris-datepicker>\n    </div>\n    <span *ngIf=\"show()\" farrisTooltip [placement]=\"'right-top'\" [content]=\"param.description\" class=\"f-icon f-icon-info\"></span>\n  </div>\n</div>",
                styles: [""]
            }] }
];
/** @nocollapse */
DateControlComponent.ctorParameters = () => [];
DateControlComponent.propDecorators = {
    param: [{ type: Input }],
    canEdit: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DateControlComponent.prototype.param;
    /** @type {?} */
    DateControlComponent.prototype.canEdit;
}
