// @farris/devkit/esm2015/lib/entity/validator/validation_types.js
const validationMessage = {};
/**
 * 验证类型定义
 */
export class ValidationTypes {
    /**
     * 判断验证类型是否合法
     * @param type 验证类型
     */
    static isValidType(type) {
        return (type !== `isValidType` &&
            type !== 'getMessage' &&
            Object.keys(this)
                .map(key => this[key])
                .indexOf(type) !== -1);
    }
    /**
     * 获取指定验证类型的提示信息
     * @param type 验证类型
     * @returns 返回验证信息
     */
    static getMessage(type) {
        const messages = validationMessage[this.CURRENT_LANGUAGE] || validationMessage['zh-CHS'];
        return messages[type] || '';
    }
    static setCurrentLanguage(language) {
        this.CURRENT_LANGUAGE = language;
    }
}
ValidationTypes.CURRENT_LANGUAGE = 'zh-CNS';
/**
 * 条件验证表达式
 */
ValidationTypes.CUSTOM_VALIDATION = `customValidation`;
/**
 * 必填
 */
ValidationTypes.REQUIRED = 'required';
/**
 * 比较值是否相同
 */
ValidationTypes.EQUALS = 'equals';
/**
 * 比较值是否不相同
 */
ValidationTypes.NOT_EQUALS = 'notEquals';
/**
 * 数字
 */
ValidationTypes.IS_NUMBER = 'isNumber';
/**
 * 整数
 */
ValidationTypes.IS_INT = 'isInt';
/**
 * 浮点数
 */
ValidationTypes.IS_FLOAT = 'isFloat';
/**
 * 字符串
 */
ValidationTypes.IS_STRING = 'isString';
/**
 * 布尔值
 */
ValidationTypes.IS_BOOLEAN = 'isBoolean';
/**
 * 日期
 */
ValidationTypes.IS_DATE = 'isDate';
/**
 * 日期字符串
 */
ValidationTypes.IS_DATE_STRING = 'isDateString';
/**
 * 布尔值字符串
 */
ValidationTypes.IS_BOOLEAN_STRING = 'isBooleanString';
/**
 * 数字字符串
 */
ValidationTypes.IS_NUMBER_STRING = 'isNumberString';
/**
 * 邮箱地址
 */
ValidationTypes.IS_EMAIL = 'isEmail';
/**
 * JSON
 */
ValidationTypes.IS_JSON = 'isJSON';
/**
 * 字符串小写
 */
ValidationTypes.IS_LOWERCASE = 'isLowercase';
/**
 * 字符串大写
 */
ValidationTypes.IS_UPPERCASE = 'isUppercase';
/**
 * 数字范围
 */
ValidationTypes.RANGE = 'range';
/**
 * 最小值
 */
ValidationTypes.MIN = 'min';
/**
 * 最小值
 */
ValidationTypes.MINVALUE = 'minValue';
/**
 * 最大值
 */
ValidationTypes.MAX = 'max';
/**
 * 最大值
 */
ValidationTypes.MAXVALUE = 'maxValue';
/**
 * 长度
 */
ValidationTypes.LENGTH = 'length';
/**
 * 字符串最大长度
 */
ValidationTypes.MAX_LENGTH = 'maxLength';
/**
 * 字符串最小长度
 */
ValidationTypes.MIN_LENGTH = 'minLength';
/**
 * 最小日期
 */
ValidationTypes.MIN_DATE = 'minDate';
/**
 * 最大日期
 */
ValidationTypes.MAX_DATE = 'maxDate';
/**
 * 排除非法字符
 */
ValidationTypes.EXCLUDE = 'exclude';
/**
 * 正则表达式
 */
ValidationTypes.MATCHES = 'matches';
ValidationTypes.FIELD_CONTAINER = 'fieldContainer';
validationMessage['zh-CHS'] = {
    fieldContainer: `$target 第 $value 行`,
    required: `请输入'$property'`,
    equals: `'$property'的值与$constraint1不相等`,
    notEquals: `'$property'的值不能与'$constraint1'相同`,
    isNumber: `'$property'的值不是数字`,
    isInt: `'$property'的值不是整数`,
    isFloat: `'$property'的值不是浮点型数字`,
    isBoolean: `'$property'的值不是布尔值`,
    isDate: `'$property'的值不是有效日期`,
    isEmail: `邮箱地址不正确`,
    min: `'$property'的值不应小于$constraint1`,
    minValue: `'$property'的值不应小于$constraint1`,
    minDate: `'$property'的日期不应早于$constraint1`,
    max: `'$property'的值不应大于$constraint1`,
    maxValue: `'$property'的值不应大于$constraint1`,
    maxDate: `'$property'不应晚于$constraint1`,
    isBooleanString: `'$property'的值不是有效布尔值`,
    isDateString: `'$property'的值不是有效的日期`,
    isLowercase: `'$property'的值应全部为小写字符串`,
    isUppercase: `'$property'的值应全部为大写字符串`,
    length: `'$property'的长度应介于$constraint1~$constraint2之间`,
    range: `'$property'的值应介于$constraint1~$constraint2之间`,
    maxLength: `'$property'的长度不得大于$constraint1`,
    minLength: `'$property'的长度不得小于$constraint1`,
    isNumberString: `'$property'的值不是数字`,
    exclude: `'$property'的值不能包含：$constraint1`,
    matches: `'$property'校验不通过`
};
validationMessage['en'] = {
    fieldContainer: `$target row $value`,
    required: `Please input '$property'`,
    equals: `'$property' should equals '$constraint1'`,
    notEquals: `'$property' should not equals '$constraint1'`,
    isNumber: `'$property' should be number`,
    isInt: `'$property' should be integer`,
    isFloat: `'$property' should be float`,
    isBoolean: `'$property' should be boolean`,
    isDate: `'$property' should be date`,
    isEmail: `'$property' should be e-mail address`,
    min: `'$property' should not less than $constraint1`,
    minValue: `'$property' should not less than $constraint1`,
    minDate: `'$property' should not early than $constraint1`,
    max: `'$property' should not bigger than $constraint1`,
    maxValue: `'$property' should not bigger than $constraint1`,
    maxDate: `'$property' should not late than $constraint1`,
    isBooleanString: `'$property' should be boolean string`,
    isDateString: `'$property' should be date string`,
    isLowercase: `'$property' should be lowercase charactor`,
    isUppercase: `'$property' should be uppercase charactor`,
    length: `'$property' length should between $constraint1~$constraint2之间`,
    range: `'$property' value should between $constraint1~$constraint2之间`,
    maxLength: `'$property' should not longer than $constraint1`,
    minLength: `'$property' should not shorter than $constraint1`,
    isNumberString: `'$property' should be number string`,
    exclude: `'$property' should not include $constraint1`,
    matches: `'$property' calibration failed`
};
validationMessage['zh-CHT'] = {
    fieldContainer: `$target 第 $value 行`,
    required: `請輸入'$property'`,
    equals: `'$property'的值與$constraint1不相等`,
    notEquals: `'$property'的值不能與'$constraint1'相同`,
    isNumber: `'$property'的值不是數字`,
    isInt: `'$property'的值不是整數`,
    isFloat: `'$property'的值不是浮點型數字`,
    isBoolean: `'$property'的值不是佈爾值`,
    isDate: `'$property'的值不是有效日期`,
    isEmail: `郵箱地址不正確`,
    min: `'$property'的值不應小於$constraint1`,
    minValue: `'$property'的值不應小於$constraint1`,
    minDate: `'$property'的日期不應早於$constraint1`,
    max: `'$property'的值不應大於$constraint1`,
    maxValue: `'$property'的值不應大於$constraint1`,
    maxDate: `'$property'不應晚於$constraint1`,
    isBooleanString: `'$property'的值不是有效佈爾值`,
    isDateString: `'$property'的值不是有效的日期`,
    isLowercase: `'$property'的值應全部爲小冩字符串`,
    isUppercase: `'$property'的值應全部爲大冩字符串`,
    length: `'$property'的長度應介於$constraint1~$constraint2之間`,
    range: `'$property'的值應介於$constraint1~$constraint2之間`,
    maxLength: `'$property'的長度不得大於$constraint1`,
    minLength: `'$property'的長度不得小於$constraint1`,
    isNumberString: `'$property'的值不是數字`,
    exclude: `'$property'的值不能包含：$constraint1`,
    matches: `'$property'校驗不通過`
};
