// @ecp-caf/caf-common/fesm5/ecp-caf-caf-common.js
import { RouterModule } from '@angular/router';
import { CommonModule } from '@angular/common';
import { __extends } from 'tslib';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { NgModule, Injectable, Optional, defineInjectable } from '@angular/core';
import { CookieService } from 'ngx-cookie-service';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CommonSharedModule = /** @class */ (function () {
    function CommonSharedModule() {
    }
    CommonSharedModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        RouterModule
                    ],
                    declarations: [],
                    exports: [],
                    providers: [],
                    entryComponents: []
                },] }
    ];
    return CommonSharedModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Abstract cache storage
 * @abstract
 */
var  /**
 * Abstract cache storage
 * @abstract
 */
CacheStorageAbstract = /** @class */ (function () {
    function CacheStorageAbstract() {
    }
    return CacheStorageAbstract;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Service for storing data in session storage
 */
var CacheSessionStorage = /** @class */ (function (_super) {
    __extends(CacheSessionStorage, _super);
    function CacheSessionStorage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    CacheSessionStorage.prototype.getItem = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var value = sessionStorage.getItem(key);
        return value ? JSON.parse(value) : null;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    CacheSessionStorage.prototype.setItem = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        try {
            sessionStorage.setItem(key, JSON.stringify(value));
            return true;
        }
        catch (e) {
            return false;
        }
    };
    /**
     * @param {?} key
     * @return {?}
     */
    CacheSessionStorage.prototype.removeItem = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        sessionStorage.removeItem(key);
    };
    /**
     * @return {?}
     */
    CacheSessionStorage.prototype.clear = /**
     * @return {?}
     */
    function () {
        sessionStorage.clear();
    };
    /**
     * @return {?}
     */
    CacheSessionStorage.prototype.type = /**
     * @return {?}
     */
    function () {
        return 1 /* SESSION_STORAGE */;
    };
    /**
     * @return {?}
     */
    CacheSessionStorage.prototype.isEnabled = /**
     * @return {?}
     */
    function () {
        try {
            sessionStorage.setItem('test', 'test');
            sessionStorage.removeItem('test');
            return true;
        }
        catch (e) {
            return false;
        }
    };
    CacheSessionStorage.decorators = [
        { type: Injectable }
    ];
    return CacheSessionStorage;
}(CacheStorageAbstract));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Service for storing data in local storage
 */
var CacheLocalStorage = /** @class */ (function (_super) {
    __extends(CacheLocalStorage, _super);
    function CacheLocalStorage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    CacheLocalStorage.prototype.getItem = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var value = localStorage.getItem(key);
        return value ? JSON.parse(value) : null;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    CacheLocalStorage.prototype.setItem = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        try {
            localStorage.setItem(key, JSON.stringify(value));
            return true;
        }
        catch (e) {
            return false;
        }
    };
    /**
     * @param {?} key
     * @return {?}
     */
    CacheLocalStorage.prototype.removeItem = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        localStorage.removeItem(key);
    };
    /**
     * @return {?}
     */
    CacheLocalStorage.prototype.clear = /**
     * @return {?}
     */
    function () {
        localStorage.clear();
    };
    /**
     * @return {?}
     */
    CacheLocalStorage.prototype.type = /**
     * @return {?}
     */
    function () {
        return 0 /* LOCAL_STORAGE */;
    };
    /**
     * @return {?}
     */
    CacheLocalStorage.prototype.isEnabled = /**
     * @return {?}
     */
    function () {
        try {
            localStorage.setItem('test', 'test');
            localStorage.removeItem('test');
            return true;
        }
        catch (e) {
            return false;
        }
    };
    CacheLocalStorage.decorators = [
        { type: Injectable }
    ];
    return CacheLocalStorage;
}(CacheStorageAbstract));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var WindowVariableService = /** @class */ (function () {
    function WindowVariableService() {
    }
    /**
     * 从当前window遍历到框架顶级window
     * 并获取一个非空的自定义变量
     */
    /**
     * 从当前window遍历到框架顶级window
     * 并获取一个非空的自定义变量
     * @param {?} key
     * @return {?}
     */
    WindowVariableService.prototype.getVariableFromWindow = /**
     * 从当前window遍历到框架顶级window
     * 并获取一个非空的自定义变量
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var topWin = this.getTopWindow();
        /** @type {?} */
        var frmWin = window;
        while (!frmWin[key] && frmWin !== topWin) {
            frmWin = frmWin.parent;
        }
        return frmWin[key];
    };
    /**
     * 从当前window遍历到框架顶级window
     * 并获取一个非空且含有元素的自定义数组变量
     */
    /**
     * 从当前window遍历到框架顶级window
     * 并获取一个非空且含有元素的自定义数组变量
     * @param {?} key
     * @return {?}
     */
    WindowVariableService.prototype.getArrayFromWindow = /**
     * 从当前window遍历到框架顶级window
     * 并获取一个非空且含有元素的自定义数组变量
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var topWin = this.getTopWindow();
        /** @type {?} */
        var frmWin = window;
        while ((!frmWin[key] || frmWin[key].length === 0) && frmWin !== topWin) {
            frmWin = frmWin.parent;
        }
        return (/** @type {?} */ (frmWin[key])) || [];
    };
    /**
     * 获取框架的顶级window
     * 进行跨域封装
     */
    /**
     * 获取框架的顶级window
     * 进行跨域封装
     * @return {?}
     */
    WindowVariableService.prototype.getTopWindow = /**
     * 获取框架的顶级window
     * 进行跨域封装
     * @return {?}
     */
    function () {
        /** @type {?} */
        var curWin = window.self;
        try {
            while (curWin.location.origin === curWin.parent.location.origin) {
                if (curWin !== curWin.parent && !((/** @type {?} */ (curWin))).isRTFTopWin) {
                    curWin = curWin.parent;
                }
                else {
                    break;
                }
            }
        }
        catch (e) { }
        return curWin;
    };
    WindowVariableService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    WindowVariableService.ctorParameters = function () { return []; };
    return WindowVariableService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Service for storing data in local storage
 */
var CacheMemoryStorage = /** @class */ (function (_super) {
    __extends(CacheMemoryStorage, _super);
    function CacheMemoryStorage() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._data = {};
        _this.windowSvc = new WindowVariableService();
        return _this;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    CacheMemoryStorage.prototype.getItem = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var cache = this.getGlobleCache(key);
        if (cache) {
            this._data[key] = cache;
        }
        return this._data[key] ? this._data[key] : null;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    CacheMemoryStorage.prototype.setItem = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        this.setGlobleCache(key, value);
        this._data[key] = value;
        return true;
    };
    /**
     * @param {?} key
     * @return {?}
     */
    CacheMemoryStorage.prototype.removeItem = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        delete this._data[key];
    };
    /**
     * @return {?}
     */
    CacheMemoryStorage.prototype.clear = /**
     * @return {?}
     */
    function () {
        this._data = [];
    };
    /**
     * @return {?}
     */
    CacheMemoryStorage.prototype.type = /**
     * @return {?}
     */
    function () {
        return 2 /* MEMORY */;
    };
    /**
     * @return {?}
     */
    CacheMemoryStorage.prototype.isEnabled = /**
     * @return {?}
     */
    function () {
        return true;
    };
    /**
     * @private
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    CacheMemoryStorage.prototype.setGlobleCache = /**
     * @private
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        /** @type {?} */
        var topWindow = this.windowSvc.getTopWindow();
        try {
            if (topWindow.name) {
                /** @type {?} */
                var exist = false;
                /** @type {?} */
                var ary = topWindow.name.split('&|&');
                // tslint:disable-next-line: prefer-for-of
                for (var i = 0; i < ary.length; i++) {
                    if (ary[i]) {
                        /** @type {?} */
                        var obj = JSON.parse(ary[i]);
                        if (obj[key]) {
                            topWindow.name = topWindow.name.replace(ary[i], this.serializeData(key, value));
                            exist = true;
                            break;
                        }
                    }
                }
                if (!exist) {
                    topWindow.name += '&|&' + this.serializeData(key, value);
                }
            }
            else {
                topWindow.name = this.serializeData(key, value);
            }
        }
        catch (_a) {
            topWindow.name = '';
        }
    };
    /**
     * @private
     * @param {?} key
     * @return {?}
     */
    CacheMemoryStorage.prototype.getGlobleCache = /**
     * @private
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var topWindow = this.windowSvc.getTopWindow();
        try {
            /** @type {?} */
            var result = void 0;
            if (topWindow.name) {
                if (topWindow.name.indexOf('&|&') >= 0) {
                    /** @type {?} */
                    var ary = topWindow.name.split('&|&');
                    for (var i = 0; i < ary.length; i++) {
                        if (ary[i]) {
                            /** @type {?} */
                            var obj = JSON.parse(ary[i]);
                            if (obj[key]) {
                                result = obj[key].value;
                                if (i === 0) {
                                    topWindow.name = topWindow.name.replace(ary[i] + '&|&', '');
                                }
                                else {
                                    topWindow.name = topWindow.name.replace('&|&' + ary[i], '');
                                }
                                break;
                            }
                        }
                    }
                }
                else {
                    /** @type {?} */
                    var obj = JSON.parse(topWindow.name);
                    if (obj[key]) {
                        result = obj[key].value;
                        topWindow.name = '';
                    }
                }
            }
            return result;
        }
        catch (_a) {
            topWindow.name = '';
            return null;
        }
    };
    /**
     * @private
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    CacheMemoryStorage.prototype.serializeData = /**
     * @private
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        /** @type {?} */
        var obj = {};
        obj[key] = {
            value: value
        };
        return JSON.stringify(obj);
    };
    CacheMemoryStorage.decorators = [
        { type: Injectable }
    ];
    return CacheMemoryStorage;
}(CacheStorageAbstract));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CACHE_PREFIX = 'CacheService';
/** @type {?} */
var DEFAULT_STORAGE = 2 /* MEMORY */;
/** @type {?} */
var DEFAULT_ENABLED_STORAGE = 1 /* SESSION_STORAGE */;
var CacheService = /** @class */ (function () {
    function CacheService(_storage) {
        this._storage = _storage;
        /**
         * Default cache options
         * CacheOptionsInterface
         *
         */
        this._defaultOptions = {
            expires: Number.MAX_VALUE,
            maxAge: Number.MAX_VALUE
        };
        /**
         * Cache prefix
         */
        this._prefix = CACHE_PREFIX;
        this._validateStorage();
    }
    /**
     * Set data to cache
     */
    /**
     * Set data to cache
     * @param {?} key
     * @param {?} value
     * @param {?=} options
     * @return {?}
     */
    CacheService.prototype.set = /**
     * Set data to cache
     * @param {?} key
     * @param {?} value
     * @param {?=} options
     * @return {?}
     */
    function (key, value, options) {
        /** @type {?} */
        var storageKey = this._toStorageKey(key);
        options = options ? options : this._defaultOptions;
        if (this._storage.setItem(storageKey, this._toStorageValue(value, options))) {
            if (!this._isSystemKey(key) && options.tag) {
                this._saveTag(options.tag, storageKey);
            }
            return true;
        }
        return false;
    };
    /**
     * Get data from cache
     */
    /**
     * Get data from cache
     * @param {?} key
     * @return {?}
     */
    CacheService.prototype.get = /**
     * Get data from cache
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var storageValue = this._storage.getItem(this._toStorageKey(key));
        /** @type {?} */
        var value = null;
        if (storageValue) {
            if (this._validateStorageValue(storageValue)) {
                value = storageValue.value;
            }
            else {
                this.remove(key);
            }
        }
        return value;
    };
    /**
     * Check if value exists
     */
    /**
     * Check if value exists
     * @param {?} key
     * @return {?}
     */
    CacheService.prototype.exists = /**
     * Check if value exists
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return !!this.get(key);
    };
    /**
     * Remove item from cache
     */
    /**
     * Remove item from cache
     * @param {?} key
     * @return {?}
     */
    CacheService.prototype.remove = /**
     * Remove item from cache
     * @param {?} key
     * @return {?}
     */
    function (key) {
        this._storage.removeItem(this._toStorageKey(key));
        this._removeFromTag(this._toStorageKey(key));
    };
    /**
     * Remove all from cache
     */
    /**
     * Remove all from cache
     * @return {?}
     */
    CacheService.prototype.removeAll = /**
     * Remove all from cache
     * @return {?}
     */
    function () {
        this._storage.clear();
    };
    /**
     * Get all tag data
     */
    /**
     * Get all tag data
     * @param {?} tag
     * @return {?}
     */
    CacheService.prototype.getTagData = /**
     * Get all tag data
     * @param {?} tag
     * @return {?}
     */
    function (tag) {
        var _this = this;
        /** @type {?} */
        var tags = this.get(this._tagsStorageKey()) || {};
        /** @type {?} */
        var result = {};
        if (tags[tag]) {
            tags[tag].forEach((/**
             * @param {?} key
             * @return {?}
             */
            function (key) {
                /** @type {?} */
                var data = _this.get(_this._fromStorageKey(key));
                if (data) {
                    result[_this._fromStorageKey(key)] = data;
                }
            }));
        }
        return result;
    };
    /**
     * Create a new instance of cache with needed storage
     */
    /**
     * Create a new instance of cache with needed storage
     * @param {?} type
     * @return {?}
     */
    CacheService.prototype.useStorage = /**
     * Create a new instance of cache with needed storage
     * @param {?} type
     * @return {?}
     */
    function (type) {
        /** @type {?} */
        var service = new CacheService(this._initStorage(type));
        service.setGlobalPrefix(this._getCachePrefix());
        return service;
    };
    /**
     * Remove all by tag
     */
    /**
     * Remove all by tag
     * @param {?} tag
     * @return {?}
     */
    CacheService.prototype.removeTag = /**
     * Remove all by tag
     * @param {?} tag
     * @return {?}
     */
    function (tag) {
        var _this = this;
        /** @type {?} */
        var tags = this.get(this._tagsStorageKey()) || {};
        if (tags[tag]) {
            tags[tag].forEach((/**
             * @param {?} key
             * @return {?}
             */
            function (key) {
                _this._storage.removeItem(key);
            }));
            delete tags[tag];
            this.set(this._tagsStorageKey(), tags);
        }
    };
    /**
     * Set global cache key prefix
     */
    /**
     * Set global cache key prefix
     * @param {?} prefix
     * @return {?}
     */
    CacheService.prototype.setGlobalPrefix = /**
     * Set global cache key prefix
     * @param {?} prefix
     * @return {?}
     */
    function (prefix) {
        this._prefix = prefix;
    };
    /**
     * Validate cache storage
     *
     */
    /**
     * Validate cache storage
     *
     * @private
     * @return {?}
     */
    CacheService.prototype._validateStorage = /**
     * Validate cache storage
     *
     * @private
     * @return {?}
     */
    function () {
        if (!this._storage) {
            this._storage = this._initStorage(DEFAULT_STORAGE);
        }
        if (!this._storage.isEnabled()) {
            this._storage = this._initStorage(DEFAULT_ENABLED_STORAGE);
        }
    };
    /**
     * Remove key from tags keys list
     */
    /**
     * Remove key from tags keys list
     * @private
     * @param {?} key
     * @return {?}
     */
    CacheService.prototype._removeFromTag = /**
     * Remove key from tags keys list
     * @private
     * @param {?} key
     * @return {?}
     */
    function (key) {
        // tslint:disable-next-line:prefer-const
        /** @type {?} */
        var tags = this.get(this._tagsStorageKey()) || {};
        /** @type {?} */
        var index;
        // tslint:disable-next-line:forin
        for (var tag in tags) {
            index = tags[tag].indexOf(key);
            if (index !== -1) {
                tags[tag].splice(index, 1);
                this.set(this._tagsStorageKey(), tags);
                break;
            }
        }
    };
    /**
     * Init storage by type
     */
    /**
     * Init storage by type
     * @private
     * @param {?} type
     * @return {?}
     */
    CacheService.prototype._initStorage = /**
     * Init storage by type
     * @private
     * @param {?} type
     * @return {?}
     */
    function (type) {
        /** @type {?} */
        var storage;
        switch (type) {
            case 1 /* SESSION_STORAGE */:
                storage = new CacheSessionStorage();
                break;
            case 0 /* LOCAL_STORAGE */:
                storage = new CacheLocalStorage();
                break;
            default: storage = new CacheMemoryStorage();
        }
        return storage;
    };
    /**
     * @private
     * @param {?} key
     * @return {?}
     */
    CacheService.prototype._toStorageKey = /**
     * @private
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return this._getCachePrefix() + key;
    };
    /**
     * @private
     * @param {?} key
     * @return {?}
     */
    CacheService.prototype._fromStorageKey = /**
     * @private
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return key.replace(this._getCachePrefix(), '');
    };
    /**
     * Prepare value to set to storage
     *
     */
    /**
     * Prepare value to set to storage
     *
     * @private
     * @param {?} value
     * @param {?} options
     * @return {?}
     */
    CacheService.prototype._toStorageValue = /**
     * Prepare value to set to storage
     *
     * @private
     * @param {?} value
     * @param {?} options
     * @return {?}
     */
    function (value, options) {
        return {
            value: value,
            options: this._toStorageOptions(options)
        };
    };
    /**
     * Prepare options to set to storage
     *
     */
    /**
     * Prepare options to set to storage
     *
     * @private
     * @param {?} options
     * @return {?}
     */
    CacheService.prototype._toStorageOptions = /**
     * Prepare options to set to storage
     *
     * @private
     * @param {?} options
     * @return {?}
     */
    function (options) {
        /** @type {?} */
        var storageOptions = {};
        storageOptions.expires = options.expires ? options.expires :
            (options.maxAge ? Date.now() + (options.maxAge * 1000) : this._defaultOptions.expires);
        storageOptions.maxAge = options.maxAge ? options.maxAge : this._defaultOptions.maxAge;
        return storageOptions;
    };
    /**
     * Validate storage value
     *
     */
    /**
     * Validate storage value
     *
     * @private
     * @param {?} value
     * @return {?}
     */
    CacheService.prototype._validateStorageValue = /**
     * Validate storage value
     *
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return !!value.options.expires && value.options.expires > Date.now();
    };
    /**
     * check if its system cache key
     *
     */
    /**
     * check if its system cache key
     *
     * @private
     * @param {?} key
     * @return {?}
     */
    CacheService.prototype._isSystemKey = /**
     * check if its system cache key
     *
     * @private
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return [this._tagsStorageKey()].indexOf(key) !== -1;
    };
    /**
     * Save tag to list of tags
     *
     *
     */
    /**
     * Save tag to list of tags
     *
     *
     * @private
     * @param {?} tag
     * @param {?} key
     * @return {?}
     */
    CacheService.prototype._saveTag = /**
     * Save tag to list of tags
     *
     *
     * @private
     * @param {?} tag
     * @param {?} key
     * @return {?}
     */
    function (tag, key) {
        /** @type {?} */
        var tags = this.get(this._tagsStorageKey()) || {};
        if (!tags[tag]) {
            tags[tag] = [key];
        }
        else {
            tags[tag].push(key);
        }
        this.set(this._tagsStorageKey(), tags);
    };
    /**
     * Get global cache prefix
     *
     *
     */
    /**
     * Get global cache prefix
     *
     *
     * @private
     * @return {?}
     */
    CacheService.prototype._getCachePrefix = /**
     * Get global cache prefix
     *
     *
     * @private
     * @return {?}
     */
    function () {
        return this._prefix;
    };
    /**
     * @private
     * @return {?}
     */
    CacheService.prototype._tagsStorageKey = /**
     * @private
     * @return {?}
     */
    function () {
        return 'CacheService_tags';
    };
    CacheService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CacheService.ctorParameters = function () { return [
        { type: CacheStorageAbstract, decorators: [{ type: Optional }] }
    ]; };
    return CacheService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CommonUtil = /** @class */ (function () {
    function CommonUtil() {
    }
    /**
     * @param {?} paraName
     * @return {?}
     */
    CommonUtil.getQueryString = /**
     * @param {?} paraName
     * @return {?}
     */
    function (paraName) {
        /** @type {?} */
        var reg = new RegExp('(^|&)' + paraName + '=([^&]*)(&|$)', 'i');
        /** @type {?} */
        var r = window.location.search.substr(1).match(reg);
        /** @type {?} */
        var paraValue = '';
        if (r) {
            paraValue = r[2];
        }
        return paraValue;
    };
    /**
     * @param {?} paraName
     * @return {?}
     */
    CommonUtil.getQueryStringFromHash = /**
     * @param {?} paraName
     * @return {?}
     */
    function (paraName) {
        /** @type {?} */
        var url = window.location.hash;
        /** @type {?} */
        var arrObj = url && url.split('?');
        if (arrObj && arrObj.length > 1) {
            /** @type {?} */
            var arrPara = arrObj[1].split('&');
            if (!arrPara) {
                return '';
            }
            /** @type {?} */
            var arr = void 0;
            // tslint:disable-next-line: prefer-for-of
            for (var i = 0; i < arrPara.length; i++) {
                arr = arrPara[i].split('=');
                if (arr != null && arr[0] === paraName) {
                    return arr[1];
                }
            }
            return '';
        }
        else {
            return '';
        }
    };
    /**
     * @return {?}
     */
    CommonUtil.getCookieMapFromDoc = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var cookieArr = document.cookie && document.cookie.split(';');
        /** @type {?} */
        var cookieMap = new Map();
        if (cookieArr) {
            cookieArr.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                /** @type {?} */
                var params = item.split('=');
                cookieMap.set(params[0], params[1]);
            }));
        }
        return cookieMap;
    };
    /**
     * 获取当前window 中funcId appId等公共变量
     * 返回结果格式：funcId=xx&appType=xxx&
     */
    /**
     * 获取当前window 中funcId appId等公共变量
     * 返回结果格式：funcId=xx&appType=xxx&
     * @return {?}
     */
    CommonUtil.getCommonVariables = /**
     * 获取当前window 中funcId appId等公共变量
     * 返回结果格式：funcId=xx&appType=xxx&
     * @return {?}
     */
    function () {
        return CommonUtil.getQueryStringFromHash('cvft');
    };
    /**
     * @private
     * @param {?} item
     * @return {?}
     */
    CommonUtil.matchCommonVariableKey = /**
     * @private
     * @param {?} item
     * @return {?}
     */
    function (item) {
        /** @type {?} */
        var itemArr = item.split('=');
        if (!itemArr || itemArr.length !== 2 || !itemArr[0]) {
            return false;
        }
        switch (itemArr[0].toLowerCase()) {
            case 'funcid': return true;
            case 'appid': return true;
            case 'appentrance': return true;
            case 'apptype': return true;
        }
        return false;
    };
    return CommonUtil;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SessionService = /** @class */ (function () {
    function SessionService() {
        this.SessionKey = 'session';
        this.SessionKeyForIde = 'sessionId';
        this.FuncSessionKey = 'FuncSession';
        window[this.FuncSessionKey] = window[this.FuncSessionKey] || new Map();
        this.windowService = new WindowVariableService();
    }
    /**
     * 获取用户sessionid
     */
    /**
     * 获取用户sessionid
     * @return {?}
     */
    SessionService.prototype.getUserSessionId = /**
     * 获取用户sessionid
     * @return {?}
     */
    function () {
        return localStorage.getItem(this.SessionKey) || '';
    };
    /**
     * 获取当前session 之前依赖router的方式改掉
     * 从url hash中获取query
     * N 栈为funcSession
     * J 栈为formToken
     */
    /**
     * 获取当前session 之前依赖router的方式改掉
     * 从url hash中获取query
     * N 栈为funcSession
     * J 栈为formToken
     * @param {?=} tabId
     * @return {?}
     */
    SessionService.prototype.getCurrentSeesionId = /**
     * 获取当前session 之前依赖router的方式改掉
     * 从url hash中获取query
     * N 栈为funcSession
     * J 栈为formToken
     * @param {?=} tabId
     * @return {?}
     */
    function (tabId) {
        if (!!tabId) {
            /** @type {?} */
            var tab = this.getTabArray().find((/**
             * @param {?} t
             * @return {?}
             */
            function (t) { return t.id === tabId; }));
            return tab && (tab.formToken || tab.sessionId);
        }
        /** @type {?} */
        var ssid = CommonUtil.getQueryStringFromHash('cvft');
        if (!ssid) {
            /** @type {?} */
            var pageId_1 = CommonUtil.getQueryStringFromHash('tabId');
            /** @type {?} */
            var tab = this.getTabArray().find((/**
             * @param {?} t
             * @return {?}
             */
            function (t) { return t.id === pageId_1; }));
            if (tab) {
                ssid = (tab.formToken || tab.sessionId);
            }
        }
        return ssid || this.getUserSessionId();
    };
    /***
     * 获取功能的session(功能重复打开，会存在多个)
     * (1912 框架移除funcSession，接口作废)
     */
    /**
     *
     * 获取功能的session(功能重复打开，会存在多个)
     * (1912 框架移除funcSession，接口作废)
     * @param {?} funcId
     * @return {?}
     */
    SessionService.prototype.getFuncSessionId = /**
     *
     * 获取功能的session(功能重复打开，会存在多个)
     * (1912 框架移除funcSession，接口作废)
     * @param {?} funcId
     * @return {?}
     */
    function (funcId) {
        /** @type {?} */
        var ssMap = new Map();
        /** @type {?} */
        var tabs = this.getTabArray().filter((/**
         * @param {?} t
         * @return {?}
         */
        function (t) { return t.funcId === funcId; }));
        if (!tabs || tabs.length === 0) {
            return ssMap;
        }
        // tslint:disable-next-line: prefer-for-of
        for (var i = 0; i < tabs.length; i++) {
            ssMap.set(tabs[i].tabId, tabs[i].sessionId);
        }
        return ssMap;
    };
    /**
     * 获取应用session（应用重复打开会存在多个）
     * (1912 框架移除funcSession，接口作废)
     */
    /**
     * 获取应用session（应用重复打开会存在多个）
     * (1912 框架移除funcSession，接口作废)
     * @param {?} appId
     * @param {?} appEntrance
     * @return {?}
     */
    SessionService.prototype.getAppSessionId = /**
     * 获取应用session（应用重复打开会存在多个）
     * (1912 框架移除funcSession，接口作废)
     * @param {?} appId
     * @param {?} appEntrance
     * @return {?}
     */
    function (appId, appEntrance) {
        /** @type {?} */
        var ssMap = new Map();
        /** @type {?} */
        var tabs = this.getTabArray().filter((/**
         * @param {?} t
         * @return {?}
         */
        function (t) { return t.appId === appId && t.appEntrance === appEntrance; }));
        if (!tabs || tabs.length === 0) {
            return ssMap;
        }
        // tslint:disable-next-line: prefer-for-of
        for (var i = 0; i < tabs.length; i++) {
            ssMap.set(tabs[i].tabId, tabs[i].sessionId);
        }
        return ssMap;
    };
    /**
     * 设置用户sessionid
     * 只是用于N版
     */
    /**
     * 设置用户sessionid
     * 只是用于N版
     * @param {?} ssid
     * @return {?}
     */
    SessionService.prototype.setUserSessionId = /**
     * 设置用户sessionid
     * 只是用于N版
     * @param {?} ssid
     * @return {?}
     */
    function (ssid) {
        // todo框架修改之后，要和IDE同步
        localStorage.setItem(this.SessionKey, ssid);
        localStorage.setItem(this.SessionKeyForIde, ssid);
    };
    /**
     * 缓存当前功能的sessionid
     * (1912 框架移除funcSession，接口作废)
     */
    /**
     * 缓存当前功能的sessionid
     * (1912 框架移除funcSession，接口作废)
     * @param {?} key
     * @param {?} ssid
     * @return {?}
     */
    SessionService.prototype.setFuncSessionId = /**
     * 缓存当前功能的sessionid
     * (1912 框架移除funcSession，接口作废)
     * @param {?} key
     * @param {?} ssid
     * @return {?}
     */
    function (key, ssid) {
        this.getFuncSessionCache().set(key, ssid);
    };
    /**
     * 缓存当前应用的sessionId
     * (1912 框架移除funcSession，接口作废)
     */
    /**
     * 缓存当前应用的sessionId
     * (1912 框架移除funcSession，接口作废)
     * @param {?} appId
     * @param {?} appEntrance
     * @param {?} ssid
     * @return {?}
     */
    SessionService.prototype.setAppSessionId = /**
     * 缓存当前应用的sessionId
     * (1912 框架移除funcSession，接口作废)
     * @param {?} appId
     * @param {?} appEntrance
     * @param {?} ssid
     * @return {?}
     */
    function (appId, appEntrance, ssid) {
        this.setFuncSessionId(appId + "#" + appEntrance, ssid);
    };
    /**
     * @private
     * @return {?}
     */
    SessionService.prototype.getFuncSessionCache = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var sessionValue = this.windowService.getVariableFromWindow(this.FuncSessionKey);
        return (/** @type {?} */ (sessionValue));
    };
    /**
     * @private
     * @return {?}
     */
    SessionService.prototype.getTabArray = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var iframeTabarray = 'iframeTabarray';
        /** @type {?} */
        var iframArr = this.windowService.getArrayFromWindow(iframeTabarray);
        return iframArr;
    };
    SessionService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SessionService.ctorParameters = function () { return []; };
    return SessionService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var HttpService = /** @class */ (function () {
    function HttpService(http, sessionService) {
        this.http = http;
        this.sessionService = sessionService;
    }
    /**
     * @param {?} url
     * @return {?}
     */
    HttpService.prototype.get = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return this.request('Get', url, this.setHeader({}));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    HttpService.prototype.post = /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        return this.request('Post', url, this.setHeader({ body: body }));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    HttpService.prototype.put = /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        return this.request('Put', url, this.setHeader({ body: body }));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    HttpService.prototype.delete = /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        return this.request('Delete', url, this.setHeader({ body: body }));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    HttpService.prototype.patch = /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        return this.request('Patch', url, this.setHeader({ body: body }));
    };
    /** 默认用当前sessionid */
    /**
     * 默认用当前sessionid
     * @param {?} options
     * @param {?=} bizContextId
     * @return {?}
     */
    HttpService.prototype.setHeader = /**
     * 默认用当前sessionid
     * @param {?} options
     * @param {?=} bizContextId
     * @return {?}
     */
    function (options, bizContextId) {
        options.headers = {
            'Content-Type': 'application/json'
        };
        /** @type {?} */
        var commonVariables = CommonUtil.getCommonVariables();
        if (commonVariables) {
            /** @type {?} */
            var commonVariableKey = 'X-CAF-Runtime-CommonVariable';
            options.headers[commonVariableKey] = commonVariables;
        }
        if (bizContextId) {
            /** @type {?} */
            var bizContextKey = 'X-CAF-Runtime-Context';
            options.headers[bizContextKey] = bizContextId;
        }
        return options;
    };
    /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    HttpService.prototype.request = /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    function (method, url, options) {
        if (options.body) {
            if (typeof options.body !== 'string') {
                options.body = JSON.stringify(options.body);
            }
        }
        return this.http.request(method, url, options)
            .pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            return data;
        })));
    };
    /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    HttpService.prototype.request$ = /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    function (method, url, options) {
        return this.http.request(method, url, options);
    };
    HttpService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    HttpService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: SessionService }
    ]; };
    return HttpService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var LanguageService = /** @class */ (function () {
    function LanguageService(cookieService) {
        this.cookieService = cookieService;
        this.languageKey = 'languageCode';
        this.curLangugaeKey = 'caf_web_language';
    }
    /**
     * @param {?} languageCode
     * @return {?}
     */
    LanguageService.prototype.setLanguageCode = /**
     * @param {?} languageCode
     * @return {?}
     */
    function (languageCode) {
        localStorage.setItem(this.languageKey, languageCode);
    };
    /**
     * @return {?}
     */
    LanguageService.prototype.getLanguageCode = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var curLancode = this.cookieService && this.cookieService.get(this.curLangugaeKey);
        if (!curLancode) {
            curLancode = CommonUtil.getCookieMapFromDoc().get(this.curLangugaeKey) ||
                localStorage.getItem(this.languageKey) ||
                'zh-CHS';
        }
        return curLancode;
    };
    LanguageService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    LanguageService.ctorParameters = function () { return [
        { type: CookieService }
    ]; };
    return LanguageService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var TabEntityService = /** @class */ (function () {
    function TabEntityService() {
        this.nameSpace = 'gsp-rtf-tabarr-service';
        this.tabarrayKey = 'tab-arry';
        this.windowService = new WindowVariableService();
        this.topWin = this.windowService.getTopWindow();
        this.topWin[this.nameSpace] = this.topWin[this.nameSpace] || {};
    }
    /**
     * @param {?} tabs
     * @return {?}
     */
    TabEntityService.prototype.setTabs = /**
     * @param {?} tabs
     * @return {?}
     */
    function (tabs) {
        if (!tabs || tabs.length === 0) {
            return;
        }
        /** @type {?} */
        var copy = this.deepClone(tabs);
        this.topWin[this.nameSpace][this.tabarrayKey] = copy;
    };
    /**
     * @return {?}
     */
    TabEntityService.prototype.getTabs = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var tabs = this.topWin[this.nameSpace][this.tabarrayKey] || [];
        if (!tabs || tabs.length === 0) {
            return tabs;
        }
        return this.deepClone(tabs);
    };
    /**
     * @param {?} tab
     * @return {?}
     */
    TabEntityService.prototype.setActiveTab = /**
     * @param {?} tab
     * @return {?}
     */
    function (tab) {
        /** @type {?} */
        var tabs = this.getTabs();
        /** @type {?} */
        var activeOld = tabs.find((/**
         * @param {?} t
         * @return {?}
         */
        function (t) { return t.active; }));
        if (!tab) {
            if (!!activeOld) {
                activeOld.active = false;
            }
            return;
        }
        if (tabs.length > 0) {
            /** @type {?} */
            var activeNew = tabs.find((/**
             * @param {?} t
             * @return {?}
             */
            function (t) { return t.id === tab.id; }));
            if (!activeNew) {
                activeNew = tab;
                if (!!activeOld) {
                    activeOld.active = false;
                }
                activeNew.active = true;
                tabs.push(activeNew);
            }
            else {
                if (!!activeOld) {
                    if (activeOld === activeNew) {
                        return;
                    }
                    else {
                        activeOld.active = false;
                        activeNew.active = true;
                    }
                }
                else {
                    activeNew.active = true;
                }
            }
            this.setTabs(tabs);
        }
    };
    /**
     * @return {?}
     */
    TabEntityService.prototype.getActiveTab = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var tabs = this.getTabs();
        /** @type {?} */
        var activeTab = tabs && tabs.find((/**
         * @param {?} t
         * @return {?}
         */
        function (t) { return t.active; }));
        /** @type {?} */
        var res = activeTab && {
            tabId: activeTab.id,
            appType: activeTab.appType,
            funcId: activeTab.funcId,
            appId: activeTab.appId,
            appEntrance: activeTab.appEntrance,
            name: activeTab.FuncName
        };
        return res;
    };
    /**
     * @private
     * @param {?} target
     * @return {?}
     */
    TabEntityService.prototype.deepClone = /**
     * @private
     * @param {?} target
     * @return {?}
     */
    function (target) {
        /** @type {?} */
        var result;
        if (typeof target === 'object') {
            if (Array.isArray(target)) {
                result = [];
                for (var i in target) {
                    result.push(this.deepClone(target[i]));
                }
            }
            else if (target === null) {
                result = null;
            }
            else if (target.constructor === RegExp) {
                result = target;
            }
            else {
                result = {};
                for (var i in target) {
                    result[i] = this.deepClone(target[i]);
                }
            }
        }
        else {
            result = target;
        }
        return result;
    };
    TabEntityService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TabEntityService.ctorParameters = function () { return []; };
    /** @nocollapse */ TabEntityService.ngInjectableDef = defineInjectable({ factory: function TabEntityService_Factory() { return new TabEntityService(); }, token: TabEntityService, providedIn: "root" });
    return TabEntityService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var FrmCommonModule = /** @class */ (function () {
    function FrmCommonModule() {
    }
    /**
     * @return {?}
     */
    FrmCommonModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: FrmCommonModule,
            providers: [
                CacheService,
                { provide: CacheStorageAbstract, useClass: CacheMemoryStorage },
                HttpService,
                SessionService,
                LanguageService,
                WindowVariableService,
                TabEntityService,
                CookieService
            ]
        };
    };
    FrmCommonModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [],
                    exports: [],
                    providers: [],
                    entryComponents: []
                },] }
    ];
    return FrmCommonModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { CommonUtil, CommonSharedModule, FrmCommonModule, CacheService, CacheStorageAbstract, CacheLocalStorage, CacheMemoryStorage, CacheSessionStorage, HttpService, SessionService, LanguageService, WindowVariableService, TabEntityService };
