// @farris/approval-comments/esm5/lib/approval-comment-editor/approval-comment-editor.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/approval-comment-editor/approval-comment-editor.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, EventEmitter, Output } from '@angular/core';
import { cloneDeep } from 'lodash-es';
import { ApiService } from '../../utils/api.service';
import { CommonUtil } from '../../utils/common.tool';
import { Domain } from '../../utils/domain.entity';
var ApprovalCommentEditorComponent = /** @class */ (function () {
    function ApprovalCommentEditorComponent(apiService, cdRef) {
        this.apiService = apiService;
        this.cdRef = cdRef;
        /**
         * 集成时传出的数据的规定名submitModal
         */
        this.valueChanged = new EventEmitter();
        this.approvalDomains = new Array();
        this.activityOptions = [];
    }
    Object.defineProperty(ApprovalCommentEditorComponent.prototype, "elementConfig", {
        get: /**
         * @return {?}
         */
        function () {
            return this._elementConfig;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._elementConfig = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initSetting();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
    };
    /**
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.initSetting = /**
     * @return {?}
     */
    function () {
        if (!this.elementConfig.beforeOpenModal) {
            return;
        }
        /** @type {?} */
        var result = this.elementConfig.beforeOpenModal();
        this.microAppId = result.parameters.microAppId;
        this.approvalCommentId = result.parameters.approvalCommentId;
        this.getAuditorsByMicroAppId(this.microAppId);
        this.approvalDomains = result.parameters.domains || [];
        this._approvalDomains = cloneDeep(this.approvalDomains);
        if (!this.approvalDomains.length) {
            this.addDomain();
        }
    };
    /**
     * @param {?} microAppId
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.getAuditorsByMicroAppId = /**
     * @param {?} microAppId
     * @return {?}
     */
    function (microAppId) {
        var _this = this;
        this.apiService.getAuditorsByMicroAppId(microAppId).subscribe((/**
         * @param {?} resp
         * @return {?}
         */
        function (resp) {
            if (resp) {
                /** @type {?} */
                var activityDefinedIdList_1 = [];
                [].concat(resp).map((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    if (item.Id) {
                        activityDefinedIdList_1.push({
                            value: item.Id,
                            text: item.Name
                        });
                    }
                }));
                _this.activityOptions = activityDefinedIdList_1;
                _this.approvalDomains.map((/**
                 * @param {?} domainData
                 * @return {?}
                 */
                function (domainData) {
                    if (domainData.activityDefinedIds) {
                        /** @type {?} */
                        var idArr = domainData.activityDefinedIds.split(',');
                        /** @type {?} */
                        var _idArr = tslib_1.__spread(new Set(idArr));
                        /** @type {?} */
                        var deleteArr_1 = [];
                        /** @type {?} */
                        var spareArr_1 = [];
                        _idArr.forEach((/**
                         * @param {?} item
                         * @return {?}
                         */
                        function (item) {
                            if (!!activityDefinedIdList_1.some((/**
                             * @param {?} activityDefinedId
                             * @return {?}
                             */
                            function (activityDefinedId) { return item.trim() === activityDefinedId.value; }))) {
                                spareArr_1.push(item);
                            }
                            else {
                                deleteArr_1.push(item);
                            }
                        }));
                        if (!_this.compareArray1EqualToArray2(_idArr, spareArr_1)) {
                            domainData.activityIdErrorMessage = "\u6D41\u7A0B\u8BBE\u8BA1\u5DF2\u53D8\u66F4\uFF0C\u8BF7\u66F4\u65B0\u5BA1\u6279\u610F\u89C1\u7EC4\u4EF6\u7ED1\u5B9A\u8282\u70B9";
                            domainData.activityDefinedIds = spareArr_1.join(',');
                        }
                    }
                }));
                _this._approvalDomains = cloneDeep(_this.approvalDomains);
                _this.cdRef.detectChanges();
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
        }));
    };
    // 新增域
    // 新增域
    /**
     * @param {?=} event
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.addDomain = 
    // 新增域
    /**
     * @param {?=} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var newDomain = new Domain();
        newDomain.domainId = CommonUtil.newGuid();
        /** @type {?} */
        var domainCodePrefix = '';
        if (this.approvalCommentId.length > 5) {
            domainCodePrefix = 'ac_' + this.approvalCommentId.slice(this.approvalCommentId.length - 4);
        }
        else {
            domainCodePrefix = 'ac_' + CommonUtil.newGuid().slice(0, 4);
        }
        /** @type {?} */
        var preIndexFix = 1 + this.approvalDomains.length;
        // 这里要增加去重
        newDomain.domainName = '新增域' + preIndexFix;
        newDomain.domainCode = domainCodePrefix + '_domain' + preIndexFix;
        this.approvalDomains.push(newDomain);
        this._approvalDomains = cloneDeep(this.approvalDomains);
        this.valueChanged.next({ $event: null, elementValue: this._approvalDomains });
    };
    // 删除域
    // 删除域
    /**
     * @param {?} event
     * @param {?} data
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.deleteDomain = 
    // 删除域
    /**
     * @param {?} event
     * @param {?} data
     * @return {?}
     */
    function (event, data) {
        /** @type {?} */
        var index = this.approvalDomains.findIndex((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.domainId === data.domainId; }));
        this.approvalDomains.splice(index, 1);
        this.approvalDomains = this.approvalDomains.slice();
        // this.updateApprovalDomainsInDom(this.approvalDomains);
        this._approvalDomains = cloneDeep(this.approvalDomains);
        this.valueChanged.next({ $event: null, elementValue: this._approvalDomains });
    };
    // 收折域
    // 收折域
    /**
     * @param {?} event
     * @param {?} data
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.foldDomain = 
    // 收折域
    /**
     * @param {?} event
     * @param {?} data
     * @return {?}
     */
    function (event, data) {
        /** @type {?} */
        var domainItem = this.approvalDomains.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.domainId === data.domainId; }));
        domainItem.isFold = !domainItem.isFold;
    };
    /**
     * @param {?} event
     * @param {?} index1
     * @param {?} index2
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.moveDomain = /**
     * @param {?} event
     * @param {?} index1
     * @param {?} index2
     * @return {?}
     */
    function (event, index1, index2) {
        if (index1 < 0 || index2 == this.approvalDomains.length) {
            return;
        }
        this.approvalDomains[index1] = this.approvalDomains.splice(index2, 1, this.approvalDomains[index1])[0];
        this.approvalDomains = this.approvalDomains.slice();
        // this.updateApprovalDomainsInDom(this.approvalDomains);
        this._approvalDomains = cloneDeep(this.approvalDomains);
        this.valueChanged.next({ $event: null, elementValue: this._approvalDomains });
    };
    // 设置域名称时需要判断是否名称重复
    // 设置域名称时需要判断是否名称重复
    /**
     * @param {?} domainId
     * @param {?} domainName
     * @param {?} index
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.onDomainNameBlur = 
    // 设置域名称时需要判断是否名称重复
    /**
     * @param {?} domainId
     * @param {?} domainName
     * @param {?} index
     * @return {?}
     */
    function (domainId, domainName, index) {
        if (!domainName || !domainName.trim()) {
            this.approvalDomains[index].nameErrorMessage = '域名称不能为空！';
            this.cdRef.detectChanges();
            return;
        }
        this.approvalDomains[index].nameErrorMessage = '';
        this.cdRef.detectChanges();
        /** @type {?} */
        var domainData = this.approvalDomains.find((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return data.domainId == domainId; }));
        /** @type {?} */
        var targetData = this.approvalDomains.filter((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return data.domainName.trim() == domainData.domainName; }));
        if (targetData.length > 1) {
            domainData.domainCode = ((/** @type {?} */ (this._approvalDomains.find((/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return data.domainId == domainId; }))))).domainCode;
            this.approvalDomains[index].nameErrorMessage = '域名称重复，请重新设置!';
            this.cdRef.detectChanges();
        }
        else {
            this._approvalDomains = cloneDeep(this.approvalDomains);
        }
        this.valueChanged.next({ $event: null, elementValue: this._approvalDomains });
    };
    // 设置域编号时需要判断是否编号重复
    // 设置域编号时需要判断是否编号重复
    /**
     * @param {?} domainId
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.onDomainCodeBlur = 
    // 设置域编号时需要判断是否编号重复
    /**
     * @param {?} domainId
     * @return {?}
     */
    function (domainId) {
        /** @type {?} */
        var domainData = this.approvalDomains.find((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return data.domainId == domainId; }));
        /** @type {?} */
        var regex = /^[a-zA-Z_][0-9a-zA-Z_]*$/;
        /** @type {?} */
        var isMatch = regex.test(domainData.domainCode);
        if (!isMatch) {
            // this.notifyService.warning('域编号必须以英文字母下划线开头，并且只能由英文字母、数字、下划线组成！');
            return;
        }
        /** @type {?} */
        var targetData = this.approvalDomains.filter((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return data.domainCode == domainData.domainCode; }));
        if (targetData.length > 1) {
            domainData.domainCode = ((/** @type {?} */ (this._approvalDomains.find((/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return data.domainId == domainId; }))))).domainCode;
            // this.notifyService.warning('域编号重复，请重新设置。');
        }
        else {
            this._approvalDomains = cloneDeep(this.approvalDomains);
        }
        this.valueChanged.next({ $event: null, elementValue: this._approvalDomains });
    };
    // 设置流程节点id时不能为空
    // 设置流程节点id时不能为空
    /**
     * @param {?} domainId
     * @param {?} activityDefinedIds
     * @param {?} index
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.onActivityDefinedIdsBlur = 
    // 设置流程节点id时不能为空
    /**
     * @param {?} domainId
     * @param {?} activityDefinedIds
     * @param {?} index
     * @return {?}
     */
    function (domainId, activityDefinedIds, index) {
        if (!activityDefinedIds || !activityDefinedIds.trim()) {
            this.approvalDomains[index].activityIdErrorMessage = '流程节点不能为空！';
            this.cdRef.detectChanges();
            return;
        }
        this.approvalDomains[index].activityIdErrorMessage = '';
        this.cdRef.detectChanges();
        this._approvalDomains = cloneDeep(this.approvalDomains);
        this.valueChanged.next({ $event: null, elementValue: this._approvalDomains });
    };
    /**
     * @param {?} index
     * @param {?} item
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.trackByFn = /**
     * @param {?} index
     * @param {?} item
     * @return {?}
     */
    function (index, item) {
        return item.domainId;
    };
    // 判断两个数组是否相等
    // 判断两个数组是否相等
    /**
     * @param {?} arr1
     * @param {?} arr2
     * @return {?}
     */
    ApprovalCommentEditorComponent.prototype.compareArray1EqualToArray2 = 
    // 判断两个数组是否相等
    /**
     * @param {?} arr1
     * @param {?} arr2
     * @return {?}
     */
    function (arr1, arr2) {
        if (arr1.length !== arr2.length) {
            return false;
        }
        for (var i = 0; i < arr1.length; i++) {
            if (arr1[i] !== arr2[i]) {
                return false;
            }
        }
        return true;
    };
    ApprovalCommentEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-approval-comment-editor',
                    template: "<div class=\"domain-areas\">\r\n    <div style=\"flex-direction: row;flex-wrap: wrap;\">\r\n        <div *ngFor=\"let data of approvalDomains; let i = index; trackBy: trackByFn\">\r\n            <div style=\"margin-top: 5px; border: 1px dashed #ccc;\">\r\n                <div style=\"display: flex; justify-content: space-between\">\r\n                    <div style=\"display: flex; align-items: center; margin: 5px 0 0 10px; height: 30px;\">\r\n                        <div style=\"margin-right: 5px\">\r\n                            <span class=\"f-icon f-icon-col-ascending\"\r\n                                [ngClass]=\"i == 0 ? 'cannotMoveDomain' : 'canMoveDomain'\"\r\n                                (click)=\"moveDomain($event, i - 1, i)\"></span>\r\n                        </div>\r\n                        <div>\r\n                            <span class=\"f-icon f-icon-col-descending\"\r\n                                [ngClass]=\"i == (approvalDomains.length -1) ? 'cannotMoveDomain' : 'canMoveDomain'\"\r\n                                (click)=\"moveDomain($event, i, i + 1)\"></span>\r\n                        </div>\r\n                    </div>\r\n                    <div style=\"display: flex; align-items: center; margin: 5px 10px 0 0; height: 30px;\">\r\n                        <div style=\"margin-right: 5px\">\r\n                            <span class=\"f-icon f-icon-arrow-chevron-down cursor-pointer\"\r\n                                [ngStyle]=\"{'transform': data.isFold ? 'rotate(90deg)' : 'none'}\"\r\n                                (click)=\"foldDomain($event, data)\"></span>\r\n                        </div>\r\n                        <div style=\"margin-left: 5px\">\r\n                            <span class=\"f-icon f-icon-yxs_delete cursor-pointer\"\r\n                                (click)=\"deleteDomain($event, data)\"></span>\r\n                        </div>\r\n                    </div>\r\n                </div>\r\n                <div class=\"f-form-layout farris-form farris-form-controls-inline\" id=\"data.domainId\">\r\n                    <div class=\"col-12\" style=\"margin-top: 5px\">\r\n                        <div class=\"farris-group-wrap\">\r\n                            <div class=\"form-group farris-form-group\">\r\n                                <label class=\"col-form-label\" title=\"\u57DF\u540D\u79F0\">\r\n                                    <span class=\"farris-label-text\">\u57DF\u540D\u79F0</span>\r\n                                </label>\r\n                                <div class=\"farris-input-wrap\">\r\n                                    <!--<input-group [(ngModel)]=\"data.domainName\" (clear)=\"onDomainNameBlur(data.domainId, data.domainName, i)\"\r\n                                                 (blurHandle)=\"onDomainNameBlur(data.domainId, data.domainName, i)\"></input-group>-->\r\n                                    <div class=\"f-cmp-inputgroup\">\r\n                                        <div class=\"input-group f-state-editable\">\r\n                                            <textarea class=\"form-control f-utils-fill text-left\"\r\n                                                style=\"display: block;overflow: auto\" [(ngModel)]=\"data.domainName\"\r\n                                                (blur)=\"onDomainNameBlur(data.domainId, data.domainName, i)\">\r\n                                                </textarea>\r\n                                        </div>\r\n                                    </div>\r\n                                    <div *ngIf=\"data?.nameErrorMessage\" style=\"margin-top: 4px\">\r\n                                        <div class=\"error-text\">\r\n                                            {{data?.nameErrorMessage}}\r\n                                        </div>\r\n                                    </div>\r\n                                </div>\r\n                            </div>\r\n                        </div>\r\n                    </div>\r\n                    <!-- <div class=\"col-12\" style=\"margin-top: 5px\" *ngIf=\"!data.isFold\">\r\n                        <div class=\"farris-group-wrap\">\r\n                            <div class=\"form-group farris-form-group\">\r\n                                <label class=\"col-form-label\" title=\"\u57DF\u7F16\u53F7\">\r\n                                    <span class=\"farris-label-text\">\u57DF\u7F16\u53F7</span>\r\n                                </label>\r\n                                <div class=\"farris-input-wrap\">\r\n                                    <input-group [(ngModel)]=\"data.domainCode\"\r\n                                                 (blurHandle)=\"onDomainCodeBlur(data.domainId)\"></input-group>\r\n                                </div>\r\n                            </div>\r\n                        </div>\r\n                    </div>-->\r\n                    <div class=\"col-12\" style=\"margin-top: 5px\" *ngIf=\"!data.isFold\">\r\n                        <div class=\"farris-group-wrap\">\r\n                            <div class=\"form-group farris-form-group\">\r\n                                <label class=\"col-form-label\" title=\"\u6D41\u7A0B\u8282\u70B9\">\r\n                                    <span class=\"farris-label-text\">\u6D41\u7A0B\u8282\u70B9</span>\r\n                                </label>\r\n                                <div class=\"farris-input-wrap\">\r\n                                    <!-- <input-group [(ngModel)]=\"data.activityDefinedIds\"\r\n                                                 (blurHandle)=\"onActivityDefinedIdsBlur(data.domainId)\"></input-group>-->\r\n                                    <farris-combo-list [(ngModel)]=\"data.activityDefinedIds\" [idField]=\"'value'\"\r\n                                        [textField]=\"'text'\" [data]=\"activityOptions\" [multiSelect]=\"true\"\r\n                                        (ngModelChange)=\"onActivityDefinedIdsBlur(data.domainId, data.activityDefinedIds, i)\"\r\n                                        [editable]=\"false\" [enableClear]=\"false\" [enableCancelSelected]=\"false\">\r\n                                    </farris-combo-list>\r\n                                    <div *ngIf=\"data?.activityIdErrorMessage\" style=\"margin-top: 4px\">\r\n                                        <div class=\"error-text\">\r\n                                            {{data?.activityIdErrorMessage}}\r\n                                        </div>\r\n                                    </div>\r\n                                </div>\r\n                            </div>\r\n                        </div>\r\n                    </div>\r\n                </div>\r\n            </div>\r\n        </div>\r\n    </div>\r\n    <div style=\"display: flex; justify-content:center;align-items: center; margin-top: 15px;\">\r\n        <input type=\"button\" class=\"btn btn-primary\" value=\"\u65B0\u589E\u57DF\" (click)=\"addDomain($event)\" />\r\n    </div>\r\n</div>",
                    styles: ["@media screen and (min-width:888px){::ng-deep .apporal-comment .farris-form .farris-group-wrap{max-width:38.625rem!important}}.domain-areas .farris-form .farris-group-wrap::after{display:none}.domain-areas .canMoveDomain{cursor:pointer}.domain-areas .cannotMoveDomain{cursor:not-allowed;color:#80808080}.farris-form-controls-inline .farris-group-wrap .col-form-label,.farris-form-inline .col-form-label{width:4rem}.error-text{color:red}"]
                }] }
    ];
    /** @nocollapse */
    ApprovalCommentEditorComponent.ctorParameters = function () { return [
        { type: ApiService },
        { type: ChangeDetectorRef }
    ]; };
    ApprovalCommentEditorComponent.propDecorators = {
        valueChanged: [{ type: Output }]
    };
    return ApprovalCommentEditorComponent;
}());
export { ApprovalCommentEditorComponent };
if (false) {
    /**
     * 集成时传出的数据的规定名submitModal
     * @type {?}
     */
    ApprovalCommentEditorComponent.prototype.valueChanged;
    /**
     * @type {?}
     * @private
     */
    ApprovalCommentEditorComponent.prototype._elementConfig;
    /** @type {?} */
    ApprovalCommentEditorComponent.prototype.approvalCommentId;
    /** @type {?} */
    ApprovalCommentEditorComponent.prototype.approvalDomains;
    /** @type {?} */
    ApprovalCommentEditorComponent.prototype._approvalDomains;
    /**
     * 记录初始入参
     * @type {?}
     */
    ApprovalCommentEditorComponent.prototype.originalData;
    /** @type {?} */
    ApprovalCommentEditorComponent.prototype.microAppId;
    /** @type {?} */
    ApprovalCommentEditorComponent.prototype.activityDefinedIds;
    /** @type {?} */
    ApprovalCommentEditorComponent.prototype.activityOptions;
    /**
     * @type {?}
     * @private
     */
    ApprovalCommentEditorComponent.prototype.apiService;
    /**
     * @type {?}
     * @private
     */
    ApprovalCommentEditorComponent.prototype.cdRef;
}
