// @gsp-wf/wf-bizprocess-lookup/esm5/lib/wf-bizprocess-lookup.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild } from '@angular/core';
import { of } from 'rxjs';
import { LookupGridComponent, ServerSideToken } from '@farris/ui-lookup';
import { WFBizProcessLookUpUIState } from './service/wf-bizprocess-lookup.uistate';
import { WFBizProcessLookUpService } from './service/wf-bizprocess-lookup.service';
var WfBizprocessLookupComponent = /** @class */ (function () {
    function WfBizprocessLookupComponent(uiState) {
        var _this = this;
        this.uiState = uiState;
        /**
         * 宽度
         */
        this.width = 550;
        /**
         * 取数源
         */
        this.sourceType = 'rtProcess';
        /**
         * 只读
         */
        this.disabled = false;
        /**
         * 是否单选
         */
        this.singleSelect = true;
        /**
         * 是否分页，默认true
         */
        this.pagination = true;
        /**
         * 首次加载第几页，默认1
         */
        this.pageIndex = 1;
        /**
         * 每页加载数，默认20
         */
        this.pageSize = 20;
        // @Input() beforeClose: (selectData: any) => Observable<PickedResult>;
        /**
         * 帮助选择后事件
         */
        this.afterConfirm = new EventEmitter();
        this.dialogClosed = new EventEmitter();
        this.clear = new EventEmitter();
        this.columns = [];
        /**
         * 帮助前事件,强行传递过滤条件
         */
        this.beforeOpen = (/**
         * @param {?} arg
         * @return {?}
         */
        function (arg) {
            // if (!this.filter) {
            //   // 可能调用方不传递，识别为非空
            //   this.filter = {};
            // }
            // this.filter.pageIndex = this.pageIndex;
            // this.filter.pageSize = this.pageSize;
            // 把过滤数据和当前的数据传递到service层
            return of({
                showDialog: true, data: {
                    filter: _this.filter
                }
            });
        });
        this.dictPicked = (/**
         * @param {?} arg
         * @return {?}
         */
        function (arg) {
            _this.afterConfirm.next(arg);
            return of({ closeDialog: true });
        });
    }
    /**
     * @return {?}
     */
    WfBizprocessLookupComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        switch (this.sourceType) {
            case 'rtProcess':
                this.columns = [
                    { title: 'Key值', field: 'id', width: '80', searchField: 'id', },
                    { title: '名称', field: 'name', width: '100', searchField: 'name', }
                ];
                this.dataUri = this.uiState.rtProcessUri;
                break;
            case 'flowform':
                this.columns = [
                    { title: '编号', field: 'code', width: '80', searchField: 'code', },
                    { title: '名称', field: 'name', width: '100', searchField: 'name', }
                ];
                this.dataUri = this.uiState.flowFormUri;
                break;
            default:
                break;
        }
    };
    /**
     * @param {?} arg
     * @return {?}
     */
    WfBizprocessLookupComponent.prototype.afterClear = /**
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        this.clear.next(arg);
    };
    /** 手动显示帮助 */
    /**
     * 手动显示帮助
     * @return {?}
     */
    WfBizprocessLookupComponent.prototype.showHelp = /**
     * 手动显示帮助
     * @return {?}
     */
    function () {
        this.lookUp.showDialog();
    };
    WfBizprocessLookupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-bizprocess-lookup',
                    template: "\n    <farris-lookup-grid #lookup\n      [disabled]=\"disabled\"\n      [editable]=\"false\"\n      [bindingData]=\"bindingData\"\n      [(ngModel)]=\"displayTxt\"\n      [columns]=\"columns\"\n      [singleSelect]=\"singleSelect\"\n      [title]=\"title\"\n      [displayType]=\"'List'\"\n      [uri]=\"dataUri\"\n      [idField]=\"'id'\"\n      [pageIndex]=\"pageIndex\"\n      [pageSize]=\"pageSize\"\n      [pagination]=\"pagination\"\n      [valueField]=\"'id'\"\n      [textField]=\"'name'\"\n      (clear)=\"afterClear($event)\"\n      [dialogWidth]=\"width\"\n      [dictPicking]=\"beforeOpen\"\n      [dictPicked]=\"dictPicked\">\n    </farris-lookup-grid>\n  ",
                    providers: [{ provide: ServerSideToken, useClass: WFBizProcessLookUpService }]
                }] }
    ];
    /** @nocollapse */
    WfBizprocessLookupComponent.ctorParameters = function () { return [
        { type: WFBizProcessLookUpUIState }
    ]; };
    WfBizprocessLookupComponent.propDecorators = {
        title: [{ type: Input }],
        width: [{ type: Input }],
        sourceType: [{ type: Input }],
        disabled: [{ type: Input }],
        singleSelect: [{ type: Input }],
        pagination: [{ type: Input }],
        pageIndex: [{ type: Input }],
        pageSize: [{ type: Input }],
        displayTxt: [{ type: Input }],
        bindingData: [{ type: Input }],
        filter: [{ type: Input }],
        afterConfirm: [{ type: Output }],
        dialogClosed: [{ type: Output }],
        clear: [{ type: Output }],
        lookUp: [{ type: ViewChild, args: ['lookup',] }]
    };
    return WfBizprocessLookupComponent;
}());
export { WfBizprocessLookupComponent };
if (false) {
    /**
     * 标题
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.title;
    /**
     * 宽度
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.width;
    /**
     * 取数源
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.sourceType;
    /**
     * 只读
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.disabled;
    /**
     * 是否单选
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.singleSelect;
    /**
     * 是否分页，默认true
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.pagination;
    /**
     * 首次加载第几页，默认1
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.pageIndex;
    /**
     * 每页加载数，默认20
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.pageSize;
    /**
     * 绑定到帮助上的数据，初始显示值用
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.displayTxt;
    /**
     * 存储帮助选择的对象
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.bindingData;
    /**
     * 过滤条件
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.filter;
    /**
     * 帮助选择后事件
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.afterConfirm;
    /** @type {?} */
    WfBizprocessLookupComponent.prototype.dialogClosed;
    /** @type {?} */
    WfBizprocessLookupComponent.prototype.clear;
    /** @type {?} */
    WfBizprocessLookupComponent.prototype.dataUri;
    /** @type {?} */
    WfBizprocessLookupComponent.prototype.columns;
    /** @type {?} */
    WfBizprocessLookupComponent.prototype.lookUp;
    /**
     * 帮助前事件,强行传递过滤条件
     * @type {?}
     */
    WfBizprocessLookupComponent.prototype.beforeOpen;
    /** @type {?} */
    WfBizprocessLookupComponent.prototype.dictPicked;
    /**
     * @type {?}
     * @private
     */
    WfBizprocessLookupComponent.prototype.uiState;
}
