// @gsp-wf/wf-process-editor/esm2015/lib/views/approvel-items/entities/approval-item.entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class ApprovalItemEntity {
    /**
     * @param {?=} value
     * @param {?=} text
     * @param {?=} wfAction
     */
    constructor(value, text, wfAction) {
        this.isDefault = false;
        this.commentNecessary = false;
        this.certification = false;
        this.attachmentNecessary = false;
        this.signatureNecessary = false;
        this.value = value;
        this.text = text;
        this.workFlowAction = wfAction;
    }
}
if (false) {
    /** @type {?} */
    ApprovalItemEntity.prototype.value;
    /** @type {?} */
    ApprovalItemEntity.prototype.text;
    /** @type {?} */
    ApprovalItemEntity.prototype.workFlowAction;
    /** @type {?} */
    ApprovalItemEntity.prototype.isDefault;
    /** @type {?} */
    ApprovalItemEntity.prototype.commentNecessary;
    /** @type {?} */
    ApprovalItemEntity.prototype.certification;
    /** @type {?} */
    ApprovalItemEntity.prototype.attachmentNecessary;
    /** @type {?} */
    ApprovalItemEntity.prototype.signatureNecessary;
    /** @type {?} */
    ApprovalItemEntity.prototype.defaultComment;
}
