// @farris/ui-common/esm2015/lib/directives/farris-template.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef } from '@angular/core';
export class FarrisTemplateDirective {
    /**
     * @param {?} template
     */
    constructor(template) {
        this.template = template;
    }
    /**
     * @return {?}
     */
    getType() {
        return this.name;
    }
}
FarrisTemplateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farrisTemplate]'
            },] }
];
/** @nocollapse */
FarrisTemplateDirective.ctorParameters = () => [
    { type: TemplateRef }
];
FarrisTemplateDirective.propDecorators = {
    type: [{ type: Input }],
    name: [{ type: Input, args: ['farrisTemplate',] }]
};
if (false) {
    /** @type {?} */
    FarrisTemplateDirective.prototype.type;
    /** @type {?} */
    FarrisTemplateDirective.prototype.name;
    /** @type {?} */
    FarrisTemplateDirective.prototype.template;
}
