// @farris/ui-common/esm2015/lib/directives/resize-observer.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Output, ElementRef, EventEmitter, Input } from '@angular/core';
import ResizeObserver from 'resize-observer-polyfill';
/** @type {?} */
const entriesMap = new WeakMap();
const ɵ0 = /**
 * @param {?} entries
 * @return {?}
 */
entries => {
    for (const entry of entries) {
        if (entriesMap.has(entry.target)) {
            /** @type {?} */
            const comp = entriesMap.get(entry.target);
            comp._resizeCallback(entry);
        }
    }
};
/** @type {?} */
const ro = new ResizeObserver((ɵ0));
export class ResizeObserverDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        this.resize = new EventEmitter();
    }
    /**
     * @param {?} entry
     * @return {?}
     */
    _resizeCallback(entry) {
        entry['bindintClientRect'] = entry.target.getBoundingClientRect();
        this.resize.emit(entry);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        /** @type {?} */
        const target = this.target ? this.target.nativeElement : this.el.nativeElement;
        entriesMap.set(target, this);
        ro.observe(target);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        /** @type {?} */
        const target = this.el.nativeElement;
        ro.unobserve(target);
        entriesMap.delete(target);
    }
}
ResizeObserverDirective.decorators = [
    { type: Directive, args: [{
                selector: '[resizeObserver]'
            },] }
];
/** @nocollapse */
ResizeObserverDirective.ctorParameters = () => [
    { type: ElementRef }
];
ResizeObserverDirective.propDecorators = {
    target: [{ type: Input, args: ['resizeObserver',] }],
    resize: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    ResizeObserverDirective.prototype.target;
    /** @type {?} */
    ResizeObserverDirective.prototype.resize;
    /** @type {?} */
    ResizeObserverDirective.prototype.el;
}
export { ɵ0 };
