// @gsp-wf/wf-dimension-def/esm5/lib/refModule/select-element/select-element.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter } from '@angular/core';
var SelectElementComponent = /** @class */ (function () {
    function SelectElementComponent() {
        /**
         * 关闭窗口事件
         */
        this.closeModal = new EventEmitter();
        /**
         * 确认选择字段
         */
        this.confirm = new EventEmitter();
        this.progressData = {
            activeIndex: 0,
            stepMessages: [
                {
                    id: 'selectobject',
                    title: '选择节点',
                    icon: ''
                },
                {
                    id: 'selectelement',
                    title: '选择字段',
                    icon: ''
                }
            ]
        };
        this.changeObj = true;
        /**
         * 字段选择信息
         */
        this.selectEleInfo = {
            be: null,
            // be实体
            selectedObj: null,
            // be节点
            BeFields: [],
            // Be字段列表
            selectedBeFields: [],
            // Be字段ID列表
            singleElement: false,
            // 是否单选
            assoRefElement: false // 是否包含关联带出字段
        };
        this.result = {
            be: null,
            // be实体
            selectedObj: null,
            // be节点
            selectedBeFields: [],
        };
    }
    /**
     * @return {?}
     */
    SelectElementComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * 页面状态变化时触发事件 {state:'', fromPageId:'', toPageId:''}
     * 其中state分别对应0：Finish:1：Cancel:2：Prev:3：Next:4：StepChange
     * @param value
     */
    /**
     * 页面状态变化时触发事件 {state:'', fromPageId:'', toPageId:''}
     * 其中state分别对应0：Finish:1：Cancel:2：Prev:3：Next:4：StepChange
     * @param {?} value
     * @return {?}
     */
    SelectElementComponent.prototype.stateChange = /**
     * 页面状态变化时触发事件 {state:'', fromPageId:'', toPageId:''}
     * 其中state分别对应0：Finish:1：Cancel:2：Prev:3：Next:4：StepChange
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        switch (value.state) {
            case 0:
                // 返回选择结果
                this.result.be = this.selectEleInfo.be;
                this.result.selectedObj = this.selectEleInfo.selectedObj;
                this.result.selectedBeFields = this.selectEleInfo.BeFields.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    /** @type {?} */
                    var ele = (/** @type {?} */ (item));
                    // 此处使用==判断值相等，若为===，则无法找到id相同的字段
                    // tslint:disable-next-line:triple-equals
                    return _this.selectEleInfo.selectedBeFields.findIndex((/**
                     * @param {?} fieldId
                     * @return {?}
                     */
                    function (fieldId) { return fieldId == ele.ID; })) >= 0;
                }));
                this.confirm.next(this.result);
                this.close();
                break;
            case 1:
                this.close();
                break;
            case 3:
                this.changeObj = !this.changeObj;
                break;
        }
    };
    /**
     * 关闭弹窗
     */
    /**
     * 关闭弹窗
     * @return {?}
     */
    SelectElementComponent.prototype.close = /**
     * 关闭弹窗
     * @return {?}
     */
    function () {
        this.closeModal.next();
    };
    SelectElementComponent.decorators = [
        { type: Component, args: [{
                    selector: 'select-element',
                    template: "<farris-wizard style=\"display:flex;flex:1\" [progressData]=\"progressData\" [stepClickable]=\"false\" [fill]=\"true\" [stepDirection]=\"'vertical'\" [stepPosition]=\"'PageLeft'\" (stateChange)=\"stateChange($event)\">\r\n    <wizard-page [pageId]=\"progressData.stepMessages[0].id\">\r\n        <select-object [data]=\"selectEleInfo\"></select-object>\r\n    </wizard-page>\r\n    <wizard-page [pageId]=\"progressData.stepMessages[1].id\">\r\n        <select-obj-element [data]=\"selectEleInfo\" [changeObj]=\"changeObj\"></select-obj-element>\r\n    </wizard-page>\r\n</farris-wizard>",
                    styles: [":host{display:flex;position:absolute;top:0;bottom:0;left:0;right:0}"]
                }] }
    ];
    /** @nocollapse */
    SelectElementComponent.ctorParameters = function () { return []; };
    return SelectElementComponent;
}());
export { SelectElementComponent };
if (false) {
    /**
     * 关闭窗口事件
     * @type {?}
     */
    SelectElementComponent.prototype.closeModal;
    /**
     * 确认选择字段
     * @type {?}
     */
    SelectElementComponent.prototype.confirm;
    /** @type {?} */
    SelectElementComponent.prototype.progressData;
    /** @type {?} */
    SelectElementComponent.prototype.changeObj;
    /**
     * 字段选择信息
     * @type {?}
     */
    SelectElementComponent.prototype.selectEleInfo;
    /** @type {?} */
    SelectElementComponent.prototype.result;
}
