// @farris/command-services/esm5/lib/data-services/edit-data.service.js
import { Injectable } from '@angular/core';
import { FrameContext } from '@farris/devkit';
import { EMPTY } from 'rxjs';
import { tap } from 'rxjs/operators';
import { FormLoadingService } from '../form-loading/form-loading.service';
import { LanguageService } from '../languag.service';
import { FormErrorService } from '../error/form-error.service';
import { EditStateUtil } from './utils/index';
/**
 * 数据编辑服务
 */
var EditDataService = /** @class */ (function () {
    /**
     * 构造函数
     */
    function EditDataService(frameContext) {
        this.frameContext = frameContext;
        this.repository = this.frameContext.repository;
        this.bindingData = this.frameContext.bindingData;
        this.loadingService = this.frameContext.injector.get(FormLoadingService, null);
        this.languageService = this.frameContext.injector.get(LanguageService, null);
        this.errorService = this.frameContext.injector.get(FormErrorService, null);
    }
    /**
     * 编辑（支持加锁）
     */
    EditDataService.prototype.edit = function (id) {
        var _this = this;
        id = id || this.bindingData.list.currentId;
        if (!id) {
            return EMPTY;
        }
        var loadingTimerId = this.loadingService.showLoadingWithDelay(500);
        var edit$ = this.repository.editEntityById(id);
        var result$ = edit$.pipe(tap(function () {
            _this.loadingService.hideDelayLoading(loadingTimerId);
            EditStateUtil.setEditState(_this.frameContext, id);
        }, function (error) {
            _this.loadingService.hideDelayLoading(loadingTimerId);
            _this.errorService.exception(_this.languageService.updateFailed, error);
        }));
        return result$;
    };
    EditDataService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    EditDataService.ctorParameters = function () { return [
        { type: FrameContext }
    ]; };
    return EditDataService;
}());
export { EditDataService };
