// @farris/ui-dialog/esm2015/lib/dialog.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PerfectScrollbarModule } from '@farris/ui-perfect-scrollbar';
import { AngularDraggableModule } from '@farris/ui-draggable';
import { ModalModule } from '@farris/ui-modal';
import { DialogComponent } from './dialog/dialog.component';
import { DialogService } from './dialog.service';
export class FarrisDialogModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: FarrisDialogModule,
            providers: [
                DialogService
            ]
        };
    }
}
FarrisDialogModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    DialogComponent
                ],
                imports: [
                    PerfectScrollbarModule,
                    AngularDraggableModule,
                    ModalModule,
                    CommonModule,
                ],
                exports: [
                    DialogComponent,
                    ModalModule
                ]
            },] }
];
