// @farris/mobile-command-services/esm2015/lib/history-back.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/history-back.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsBridgeService } from './js-bridge-service/index';
/**
 * 帮助映射服务
 */
export class HistoryBackService {
    constructor() {
    }
    /**
     * @return {?}
     */
    init() {
        window['MOBILE_ORIGIN_BACK_ARRAY'] = window['MOBILE_ORIGIN_BACK_ARRAY'] || [];
        window['MOBILE_ORIGIN_BACK'] = {
            /**
             * @param {?} fn
             * @return {?}
             */
            pushOriginGoback(fn) {
                window['MOBILE_ORIGIN_BACK_ARRAY'].push(fn);
            },
            /**
             * @return {?}
             */
            popOriginGoback() {
                window['MOBILE_ORIGIN_BACK_ARRAY'].pop();
            },
            /**
             * @return {?}
             */
            reflushOriginGoback() {
                window['MOBILE_ORIGIN_BACK_ARRAY'] = [];
            }
        };
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    pushOriginGoback(fn) {
        window['MOBILE_ORIGIN_BACK_ARRAY'].push(fn);
    }
    /**
     * @return {?}
     */
    popOriginGoback() {
        window['MOBILE_ORIGIN_BACK_ARRAY'].pop();
    }
    /**
     * @return {?}
     */
    reflushOriginGoback() {
        window['MOBILE_ORIGIN_BACK_ARRAY'] = [];
    }
    /**
     * @return {?}
     */
    proxyBack() {
        if (!(navigator.userAgent.toLowerCase().search('emmcloud') !== -1)) {
            return;
        }
        if (window.top.location.pathname.indexOf('mobiletaskcenter') > -1) {
            setTimeout((/**
             * @return {?}
             */
            () => {
                window.top['detailBackBtnClickHandler'] = (/**
                 * @param {?} back
                 * @return {?}
                 */
                (back) => {
                    /** @type {?} */
                    const gobackFn = window['MOBILE_ORIGIN_BACK_ARRAY'].pop();
                    if (gobackFn) {
                        gobackFn();
                        return false;
                    }
                    else {
                        /** @type {?} */
                        const key = location.pathname + location.hash.slice(0, location.hash.indexOf('?'));
                        /** @type {?} */
                        let need_go_back = sessionStorage.getItem(key);
                        if (!need_go_back) {
                            return true;
                        }
                        history.back();
                        return false;
                    }
                });
            }), 0);
        }
        else {
            /** @type {?} */
            const _close = (/**
             * @return {?}
             */
            () => {
                /** @type {?} */
                const gobackFn = window['MOBILE_ORIGIN_BACK_ARRAY'].pop();
                if (gobackFn) {
                    gobackFn();
                }
                else {
                    history.back();
                }
            });
            /** @type {?} */
            const jsBridgeService = new JsBridgeService();
            jsBridgeService.setOriginGoback(_close);
        }
    }
}
