// @gsp-cmp/querysolution/esm5/lib/utils/extend-locale.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/extend-locale.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var EXTEND_QUERYSOLUTION_LOCALE = {
    "zh-CHS": {
        extendQuerySolution: {
            private: {
                property: '个人',
                setAsPrivate: '用户个人方案'
            },
            public: {
                label: '系统公共',
                property: '系统公共',
                setAsPublic: '系统公共方案'
            },
            org: {
                label: '组织公共',
                property: '组织公共',
                setAsOrgPublic: '组织公共方案',
                placeholder: '请选择'
            },
            auth: {
                createError: '您暂无权限设置公共类型方案。',
                updateError: '您暂无权限修改公共类型方案。',
                deleteError: '您暂无权限删除公共类型方案。',
                apiError: '调用获取权限接口失败！',
                noDataAuthority: '您暂无权限设置组织公共方案。',
                noSelectedOrg: '请选择组织公共方案所属组织。'
            }
        },
        advancedMode: {
            text: {
                sysAdvancedName: '（高级）',
                standardMode: '标准模式',
                advancedMode: '高级模式',
                addCondition: '添加条件',
                addConditionGroup: '生成条件组',
                and: '且',
                or: '或',
                reset: '重置'
            },
            compareType: {
                Equal: '等于',
                NotEqual: '不等于',
                Greater: '大于',
                GreaterOrEqual: '大于等于',
                Less: '小于',
                LessOrEqual: '小于等于',
                Like: '包含',
                LikeStartWith: '开头是',
                LikeEndWith: '结尾是',
                NotLike: '不包含'
            },
            auth: {
                conditionEmptyError: '高级模式下筛选项值不应为空。'
            },
            info: {
                deleteGroup: '是否确认取消该条件组？',
                resetInfo: '是否确认重置条件？',
                deleteGroupTooltip: '取消组',
                noConditionInfo: '请先选择筛选条件。',
                existGroupInfo: '选择的条件已存在分组',
                checkedGroupInfo: '跨组操作时，应勾选该组下全部数据',
                deleteGroupInfo: '已经是最外层分组了'
            }
        }
    },
    "zh-CHT": {
        extendQuerySolution: {
            private: {
                property: '個人',
                setAsPrivate: '用戶個人方案'
            },
            public: {
                label: '系統公共',
                property: '系統公共',
                setAsPublic: '系統公共方案'
            },
            org: {
                label: '組織公共',
                property: '組織公共',
                setAsOrgPublic: '組織公共方案',
                placeholder: '請選擇'
            },
            auth: {
                createError: '您暫無權限設置公共類型方案。',
                updateError: '您暫無權限修改公共類型方案。',
                deleteError: '您暫無權限刪除公共類型方案。',
                apiError: '調用獲取權限接口失敗！',
                noDataAuthority: '您暫無權限設置组织公共方案。',
                noSelectedOrg: '請選擇組織公共方案所屬組織。'
            }
        },
        advancedMode: {
            text: {
                sysAdvancedName: '（高級）',
                standardMode: '標准模式',
                advancedMode: '高級模式',
                addCondition: '添加條件',
                addConditionGroup: '生成條件組',
                and: '且',
                or: '或',
                reset: '重置'
            },
            compareType: {
                Equal: '等于',
                NotEqual: '不等于',
                Greater: '大于',
                GreaterOrEqual: '大于等于',
                Less: '小于',
                LessOrEqual: '小于等于',
                Like: '包含',
                LikeStartWith: '開頭是',
                LikeEndWith: '結尾是',
                NotLike: '不包含'
            },
            auth: {
                conditionEmptyError: '高級模式下篩選項值不應爲空。'
            },
            info: {
                deleteGroup: '是否確認取消該條件組？',
                resetInfo: '是否確認重置條件？',
                deleteGroupTooltip: '取消組',
                noConditionInfo: '請先選擇篩選條件。',
                existGroupInfo: '選擇的條件已存在分組。',
                checkedGroupInfo: '跨組操作時，應勾選該組下全部數據。',
                deleteGroupInfo: '已經是最外層分組了。'
            }
        }
    },
    en: {
        extendQuerySolution: {
            private: {
                property: 'Private Plan',
                setAsPrivate: 'Private Plan'
            },
            public: {
                label: 'Public',
                property: 'Public Plan',
                setAsPublic: 'Public Plan'
            },
            org: {
                label: 'Org',
                property: 'Org Plan',
                setAsOrgPublic: 'Org Plan',
                placeholder: 'Please select'
            },
            auth: {
                createError: 'Donnot have authority to set public solution.',
                updateError: 'Donnot have authority to save public solution.',
                deleteError: 'Donnot have authority to delete public solution.',
                apiError: 'Failed to get authority data!',
                noDataAuthority: 'Donnot have authority to set org-public solution.',
                noSelectedOrg: 'Please select organizations for solution.'
            }
        },
        advancedMode: {
            text: {
                sysAdvancedName: '(Advanced)',
                standardMode: 'Standard Mode',
                advancedMode: 'Advanced Mode',
                addCondition: 'Add Condition',
                addConditionGroup: 'Generate Condition Group',
                and: 'And',
                or: 'Or',
                reset: 'Reset'
            },
            compareType: {
                Equal: 'Equal',
                NotEqual: 'NotEqual',
                Greater: 'Greater',
                GreaterOrEqual: 'GreaterOrEqual',
                Less: 'Less',
                LessOrEqual: 'LessOrEqual',
                Like: 'Like',
                LikeStartWith: 'LikeStartWith',
                LikeEndWith: 'LikeEndWith',
                NotLike: 'NotLike'
            },
            auth: {
                conditionEmptyError: 'Filter condition value should not be empty in advanced mode.'
            },
            info: {
                deleteGroup: 'Are you sure to cancel the condition group?',
                resetInfo: 'Are you sure to reset conditions?',
                deleteGroupTooltip: 'Cancel Condition Group.',
                noConditionInfo: 'Please select conditions first.',
                existGroupInfo: 'The selected condition already has a grouping.',
                checkedGroupInfo: 'When operating across groups, all data under that group should be checked.',
                deleteGroupInfo: 'It is already the outermost group.'
            }
        }
    }
};
