// @gsp-wf/wf-process-editor/esm2015/lib/views/activity-list/activity-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, EventEmitter, Inject, LOCALE_ID, Optional, Output, TemplateRef, ViewChild } from '@angular/core';
import { DatagridComponent } from '@farris/ui-datagrid';
import { MessagerService } from '@farris/ui-messager';
import { ProcessDesignerUIState } from '../../app/process-designer.uistate';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../i18n/pipe/wf-process-editor-locale.pipe';
export class ActivityListComponent {
    /**
     * @param {?} designerUIState
     * @param {?} message
     * @param {?} cdr
     * @param {?} localeId
     * @param {?} resources
     * @param {?} localePipe
     */
    constructor(designerUIState, message, cdr, localeId, resources, localePipe) {
        this.designerUIState = designerUIState;
        this.message = message;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.closeEmitter = new EventEmitter();
        this.submitEmitter = new EventEmitter();
        this.multiSelect = true;
        this.showCheckbox = true;
        this.data = new Array();
        this.columns = [
            { title: this.localePipe.transform('component.activityList.name'), field: 'Name', width: 60 },
            { title: 'ID', field: 'Id', width: 90 }
        ];
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.data = this.designerUIState.userActList;
    }
    /**
     * @return {?}
     */
    clickOK() {
        /** @type {?} */
        const result = this.grid.checkeds;
        if (!result || result.length === 0) {
            this.message.info(this.localePipe.transform('component.activityList.checkeds'));
            return;
        }
        this.submitEmitter.emit(result);
    }
    /**
     * @return {?}
     */
    clickCancel() {
        this.closeEmitter.emit();
    }
}
ActivityListComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-activity-list',
                template: "<div class=\"f-page\">\r\n  <div class=\"f-page-main f-utils-fill-flex-column h-100\">\r\n    <div class=\"f-utils-fill-flex-column farris-overflow-y-auto h-100\">\r\n      <farris-datagrid #grid [data]=\"data\" [columns]=\"columns\" [idField]=\"'Id'\" [fit]=\"true\" [fitColumns]=\"true\"\r\n         [pagination]=\"false\" [showCheckbox]=\"showCheckbox\" [multiSelect]=\"multiSelect\" [showLineNumber]=\"true\" lineNumberTitle=\"{{'component.activityList.checkeds' | wfLocale}}\"\r\n         [lineNumberWidth]=\"50\" [checkOnSelect]=\"true\" [selectOnCheck]=\"true\">\r\n      </farris-datagrid>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #buttonRef>\r\n    <button type=\"button\" class=\"btn btn-secondary btn-sm px-3\" (click)=\"clickCancel()\">{{'component.activityList.clickCancel' | wfLocale}}</button>\r\n    <button type=\"button\" class=\"btn btn-primary btn-sm px-3 mx-2\" (click)=\"clickOK()\">{{'component.activityList.clickOK' | wfLocale}}</button>\r\n</ng-template>\r\n",
                providers: [
                    WfLocalePipe,
                ],
                styles: [""]
            }] }
];
/** @nocollapse */
ActivityListComponent.ctorParameters = () => [
    { type: ProcessDesignerUIState },
    { type: MessagerService },
    { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
    { type: WfLocalePipe, decorators: [{ type: Optional }] }
];
ActivityListComponent.propDecorators = {
    closeEmitter: [{ type: Output }],
    submitEmitter: [{ type: Output }],
    buttonRef: [{ type: ViewChild, args: ['buttonRef',] }],
    grid: [{ type: ViewChild, args: ['grid',] }]
};
if (false) {
    /** @type {?} */
    ActivityListComponent.prototype.closeEmitter;
    /** @type {?} */
    ActivityListComponent.prototype.submitEmitter;
    /** @type {?} */
    ActivityListComponent.prototype.buttonRef;
    /** @type {?} */
    ActivityListComponent.prototype.grid;
    /** @type {?} */
    ActivityListComponent.prototype.multiSelect;
    /** @type {?} */
    ActivityListComponent.prototype.showCheckbox;
    /** @type {?} */
    ActivityListComponent.prototype.data;
    /** @type {?} */
    ActivityListComponent.prototype.columns;
    /**
     * @type {?}
     * @private
     */
    ActivityListComponent.prototype.designerUIState;
    /**
     * @type {?}
     * @private
     */
    ActivityListComponent.prototype.message;
    /**
     * @type {?}
     * @private
     */
    ActivityListComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    ActivityListComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    ActivityListComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    ActivityListComponent.prototype.localePipe;
}
