// @ecp-caf/caf-common/esm5/lib/caf.common.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CacheService } from './cache/services/cache.service';
import { CacheStorageAbstract } from './cache/services/storage/cache.storage.abstract.service';
import { CacheMemoryStorage } from './cache/services/storage/memory/cache.memory.service';
import { HttpService } from './services/httpservice/http.service';
import { SessionService } from './services/sessionservice/session.service';
import { LanguageService } from './services/languageservice/language.service';
import { WindowVariableService } from './services/window-service/window-variable.service';
import { TabEntityService } from './services/tabservice/tab-entity.service';
import { CookieService } from 'ngx-cookie-service';
var FrmCommonModule = /** @class */ (function () {
    function FrmCommonModule() {
    }
    /**
     * @return {?}
     */
    FrmCommonModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: FrmCommonModule,
            providers: [
                CacheService,
                { provide: CacheStorageAbstract, useClass: CacheMemoryStorage },
                HttpService,
                SessionService,
                LanguageService,
                WindowVariableService,
                TabEntityService,
                CookieService
            ]
        };
    };
    FrmCommonModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [],
                    exports: [],
                    providers: [],
                    entryComponents: []
                },] }
    ];
    return FrmCommonModule;
}());
export { FrmCommonModule };
