// @gsp-lcm/metadata-selector/esm2015/metadata/metadata-service.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-service.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { GSPMetadataService } from './metadata.service';
import { HttpClient } from '@angular/common/http';
import { GSP } from '@farris/ide-devkit';
/**
 * @param {?} http
 * @param {?} url
 * @return {?}
 */
export function factory(http, url) {
    return new GSPMetadataService(http, url);
}
export class GSPMetadataServiceModule {
    /**
     * @param {?} url
     * @return {?}
     */
    static forRoot(url) {
        return {
            ngModule: GSPMetadataServiceModule,
            providers: [
                { provide: 'url', useValue: url },
                { provide: GSPMetadataService, useFactory: factory, deps: [HttpClient, 'url'] }
            ]
        };
    }
}
GSPMetadataServiceModule.decorators = [
    { type: NgModule, args: [{
                exports: [],
                providers: []
            },] }
];
/** @type {?} */
const globalGsp = ((/** @type {?} */ (window.top))).gsp;
if (!((/** @type {?} */ (window))).gsp) {
    ((/** @type {?} */ (window))).gsp = /*(globalGsp && globalGsp.createChild()) ||*/ new GSP(globalGsp);
}
