// @gsp-lcm/metadata-selector/esm2015/metadata/metadata.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { MetadataPackageDto } from './metadatapackagedto';
import { Inject } from '@angular/core';
import { Subject } from 'rxjs';
export class GSPMetadataService {
    /**
     * @param {?} http
     * @param {?} url
     */
    constructor(http, url) {
        this.http = http;
        this.url = url;
        this.METADATA_BASE_PATH = '/api/dev/main/v1.0/metadatas';
        this.GSP_PROJECT_BASE_PATH = '/api/dev/main/v1.0/gsp-projects';
        this.Project_Base_Path = '/api/dev/main/v1.0/project';
        this.DBO_BASE_PATH = '/api/dev/main/v1.0/database-objects'; // dbo相关
        // dbo相关
        this.METADATA_PROJECT_BASE_PATH = '/api/dev/main/v1.0/metadata-projects';
        this.METADATA_PACKAGE_BASE_PATH = '/api/dev/main/v1.0/metadata-packages';
        this.METADATA_CONFIG_BASE_PATH = '/api/dev/main/v1.0/metadata-configs';
        // private Metadata4RefDto_Base_Path = '/api/dev/main/v1.0/metadata4refdtos'; pickMetadataIndex，dotnet不需要
        // tslint:disable-next-line:variable-name
        this.DevBasinBOInfo_Base_Path = '/api/dev/main/v1.0/dev-basic-boinfos'; // 业务对象
        // 业务对象
        this.NUGET_BASE_PATH = '/api/dev/main/v1.0/nuget-packages';
        this.SETTING_BASE_PATH = '/api/dev/main/v1.0/ide/settings';
        this.MAVEN_JAVA_BASE_PATH = '/api/dev/main/v1.0/repo-packages';
        this.FS_JAVA_BASE_PATH = '/api/dev/main/v1.0/file-system';
        this.JAVA_CODE_GENERATOR = '/api/dev/main/v1.0/generator';
        this.JAVA_BUILD = '/api/dev/main/v1.0/repo-packages';
        this.METADATA_JAVA_BASE_PATH = '/api/dev/main/v1.0/mdservice';
        this.SERVER_IP = (url === null || url === undefined) ? '' : url;
        this.sessionId = gsp.cache.get('sessionId');
    }
    /**
     * 元数据设计时服务
     * @return {?}
     */
    // 打包（功能触发）
    Pack() {
        if (this.selectType === 'folder' || this.selectType === 'project') {
            if (this.selectPath !== '' && this.selectPath !== undefined) {
                this.PackageBuild(this.selectPath.substring(1));
            }
        }
        else {
            alert('请选择合适的目录');
        }
    }
    // 打包（外部调用）
    /**
     * @param {?} path
     * @return {?}
     */
    PackageBuild(path) {
        /** @type {?} */
        const sessionId = gsp.cache.get('sessionId');
        /** @type {?} */
        const headers = new HttpHeaders({ 'Content-Type': 'application/json' }).append('SessionId', sessionId);
        /** @type {?} */
        const packageDto = new MetadataPackageDto('', '', '', path);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PACKAGE_BASE_PATH;
        return this.http.post(url, packageDto, {
            headers,
            observe: 'response'
        });
    }
    // 递归查找是否存在工程文件
    /**
     * @param {?} path
     * @return {?}
     */
    IsProjExistRecursively(path) {
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/validationrecursively?path=' + path;
        /** @type {?} */
        const metadata = new Subject();
        this.http.get(url).subscribe((/**
         * @param {?} result
         * @return {?}
         */
        result => {
            // if (result) {
            metadata.next(result);
            // }
        }));
        return metadata;
    }
    // 初始化元数据实体
    /**
     * @param {?} metadataDto
     * @return {?}
     */
    InitializeMetadataEntity(metadataDto) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        // tslint:disable-next-line:max-line-length
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_BASE_PATH + '/initialized?nameSpace=' + metadataDto.nameSpace + '&code=' + metadataDto.code + '&name=' + metadataDto.name + '&type=' + metadataDto.type + '&bizObjectID=' + metadataDto.bizobjectID + '&metadataPath=' + metadataDto.relativePath + '&extendProperty=' + metadataDto.extendProperty;
        return this.http.get(url, { headers });
    }
    // 新建元数据
    /**
     * @param {?} metadataDto
     * @return {?}
     */
    CreateMetadata(metadataDto) {
        /** @type {?} */
        const content = {
            'ID': metadataDto.id,
            'NameSpace': metadataDto.nameSpace,
            'Code': metadataDto.code,
            'Name': metadataDto.name,
            'FileName': metadataDto.fileName,
            'RelativePath': metadataDto.relativePath,
            'Content': metadataDto.content,
            'Type': metadataDto.type,
            'BizobjectID': metadataDto.bizobjectID,
            'ExtendProperty': metadataDto.extendProperty
        };
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_BASE_PATH;
        return this.http.post(url, content, {
            params: {},
            headers,
            observe: 'response'
        });
    }
    // 删除元数据文件
    /**
     * @return {?}
     */
    DeleteMetadata() {
        if (this.selectPath !== '' && this.selectPath !== undefined) {
            if (this.selectType === 'file') {
                this.DeleteMetadataOrDir(this.selectPath.substring(1)).subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                result => {
                    // 刷新节点
                    // this.treeModel.removeFileNode();
                    alert('删除成功！');
                    this.selectPath = '';
                    this.selectType = '';
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                error => {
                    alert('删除失败！');
                }));
            }
            else {
                alert('请选择正确的文件进行删除！');
            }
        }
        else {
            alert('请选择正确的文件进行删除！');
        }
    }
    // 保存元数据
    /**
     * @param {?} metadataDto
     * @param {?} sessionId
     * @return {?}
     */
    SaveMetadata(metadataDto, sessionId) {
        /** @type {?} */
        const content = {
            'ID': metadataDto.id,
            'NameSpace': metadataDto.nameSpace,
            'Code': metadataDto.code,
            'Name': metadataDto.name,
            'FileName': metadataDto.fileName,
            'RelativePath': metadataDto.relativePath,
            'Content': metadataDto.content,
            'Type': metadataDto.type,
            'BizobjectID': metadataDto.bizobjectID,
            'ExtendProperty': metadataDto.extendProperty,
            'Extendable': metadataDto.extendable
        };
        // tslint:disable-next-line:one-variable-per-declaration
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_BASE_PATH;
        return this.http.put(url, content, {
            headers,
            observe: 'response'
        });
    }
    // Load元数据（外部调用）
    /**
     * @param {?} fullName
     * @param {?} path
     * @return {?}
     */
    LoadMetadata(fullName, path) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const metadataFullPath = path.replace(/\\/g, '/') + '/' + fullName;
        /** @type {?} */
        const encMetadataFullPath = encodeURIComponent(metadataFullPath);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_BASE_PATH + '/load?metadataFullPath=' + encMetadataFullPath;
        return this.http.get(url, { headers });
    }
    // Load bo内元数据（外部调用）
    /**
     * @param {?} fullName
     * @param {?} path
     * @param {?} currentProjPath
     * @return {?}
     */
    LoadMetadataInBo(fullName, path, currentProjPath) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const metadataFullPath = path.replace(/\\/g, '/') + '/' + fullName;
        /** @type {?} */
        const encMetadataFullPath = encodeURIComponent(metadataFullPath);
        // tslint:disable-next-line: max-line-length
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_BASE_PATH + '/metadataInBo?metadataFullPath=' + encMetadataFullPath
            + '&currentProjPath=' + currentProjPath;
        return this.http.get(url, { headers });
    }
    // 删除文件夹
    /**
     * @return {?}
     */
    DeleteFolder() {
        if (this.selectPath !== '' && this.selectPath !== undefined) {
            if (this.selectType === 'folder') {
                this.DeleteMetadataOrDir(this.selectPath.substring(1)).subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                result => {
                    // 刷新节点
                    // this.treeModel.removeFolderNode();
                    alert('删除成功！');
                    this.selectPath = '';
                    this.selectType = '';
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                error => {
                    alert('删除失败！');
                }));
            }
            else {
                alert('请选择正确的目录进行删除！');
            }
        }
        else {
            alert('请选择正确的目录进行删除！');
        }
    }
    // 删除工程
    /**
     * @return {?}
     */
    DeleteProj() {
        if (this.selectPath !== '' && this.selectPath !== undefined) {
            if (this.selectType === 'project') {
                this.DeleteProject(this.selectPath.substring(1)).subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                result => {
                    // 刷新节点
                    // this.treeModel.removeProjectNode();
                    alert('删除成功！');
                    this.selectPath = '';
                    this.selectType = '';
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                error => {
                    alert('删除失败！');
                }));
            }
            else {
                alert('请选择正确的工程节点进行删除！');
            }
        }
        else {
            alert('请选择正确的工程节点进行删除！');
        }
    }
    // 元数据或者文件夹的删除（外部调用）
    /**
     * @param {?} fullPath
     * @return {?}
     */
    DeleteMetadataOrDir(fullPath) {
        /** @type {?} */
        const sessionId = gsp.cache.get('sessionId');
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', sessionId);
        /** @type {?} */
        const encFullPath = encodeURIComponent(fullPath.replace(/\\/g, '/'));
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_BASE_PATH + '/delete?fullPath=' + encFullPath;
        return this.http.delete(url, {
            params: {},
            headers,
            observe: 'response'
        });
    }
    // 判断元数据是否被依赖
    /**
     * @param {?} fullPath
     * @return {?}
     */
    isMetadataRefed(fullPath) {
        /** @type {?} */
        const sessionId = gsp.cache.get('sessionId');
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', sessionId);
        /** @type {?} */
        const encFullPath = encodeURIComponent(fullPath.replace(/\\/g, '/'));
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/isMetadataRefed?path=' + encFullPath;
        return this.http.get(url, {
            params: {},
            headers,
            observe: 'response'
        });
    }
    // 删除工程
    /**
     * @param {?} path
     * @return {?}
     */
    DeleteProject(path) {
        /** @type {?} */
        const sessionId = gsp.cache.get('sessionId');
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.Project_Base_Path + '/' + path.replace(/\\/g, '/');
        return this.http.delete(url, {
            params: {},
            headers,
            observe: 'response'
        });
    }
    /**
     * @return {?}
     */
    GetDevRootPath() {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const options = {
            params: {},
            headers,
            responseType: 'text'
        };
        /** @type {?} */
        const url = this.SERVER_IP + this.Project_Base_Path + '/rootpath';
        /** @type {?} */
        const metadata = new Subject();
        this.http.get(url, { responseType: 'text' }).subscribe((/**
         * @param {?} result
         * @return {?}
         */
        result => {
            if (result) {
                metadata.next(result);
            }
        }));
        return metadata;
    }
    /**
     * @param {?} fullName
     * @param {?} path
     * @param {?} sessionId
     * @return {?}
     */
    LoadDBO(fullName, path, sessionId) {
        /** @type {?} */
        const dboFullPath = path.replace(/\\/g, '/') + '/' + fullName;
        /** @type {?} */
        const url = this.SERVER_IP + this.DBO_BASE_PATH + '/' + dboFullPath;
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        return this.http.get(url, { headers });
    }
    // 新建文件夹
    /**
     * @param {?} path
     * @param {?} name
     * @return {?}
     */
    CreateDir(path, name) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.Project_Base_Path + '/folder' + path + '/' + name + '?quiet=true';
        return this.http.post(url, null, {
            params: {},
            headers,
            observe: 'response'
        });
    }
    // 获取元数据工程信息
    /**
     * @param {?} path
     * @return {?}
     */
    GetProjectInfo(path) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '?path=' + path;
        return this.http.get(url, { headers });
    }
    // 获取工程信息
    /**
     * @param {?} path
     * @return {?}
     */
    GetGspProjInfo(path) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.GSP_PROJECT_BASE_PATH + '?path=' + path;
        return this.http.get(url, { headers });
    }
    // 获取工程路径
    /**
     * @param {?} path
     * @return {?}
     */
    GetProjPath(path) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/path?path=' + path;
        return this.http.get(url, { responseType: 'text' });
    }
    // 获取当前工程下或者其他元数据包中依赖的元数据
    /**
     * @param {?} path
     * @param {?} metadataid
     * @param {?} sessionId
     * @return {?}
     */
    GetRefMetadata(path, metadataid, sessionId) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_BASE_PATH + '/relied?metadataPath=' + path + '&metadataID=' + metadataid;
        return this.http.get(url, { headers });
    }
    // 获取当前工程下所有的元数据包
    /**
     * @param {?} spacePath
     * @param {?} typeName
     * @return {?}
     */
    GetMetadataList(spacePath, typeName) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '?path=' + spacePath + '&metadataTypeList=' + typeName;
        return this.http.get(url, { headers });
    }
    // 获取bo工程下所有的元数据列表
    /**
     * @param {?} projPath
     * @param {?} typeName
     * @return {?}
     */
    GetMetadataListInBo(projPath, typeName) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/metadataListInBo?path=' + projPath + '&metadataTypeList=' + typeName;
        return this.http.get(url, { headers });
    }
    /**
     * @param {?} path
     * @param {?} typeCode
     * @param {?} filter
     * @param {?} postfix
     * @return {?}
     */
    GetMetadataByFilter(path, typeCode, filter, postfix) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_BASE_PATH + '/filter?path=' + path +
            '&typeCode=' + typeCode + '&filter=' + filter + '&postfix=' + postfix;
        return this.http.get(url, { headers });
    }
    // pickMetadataIndex(scope: number, currentProjPath: string, metadataIndexDto: any): Observable<any> {
    //   const sessionId = gsp.cache.get('sessionId');
    //   const headers = new HttpHeaders({ 'Content-Type': 'application/json' }).append('SessionId', sessionId);
    //   const url = this.SERVER_IP + this.Metadata4RefDto_Base_Path + '/scope/' + scope + '?currentPath=' + currentProjPath;
    //   return this.http.post(url, metadataIndexDto, {
    //     headers
    //   });
    // }
    /**
     * @param {?} scope
     * @param {?} currentProjPath
     * @param {?} metadataIndexDto
     * @return {?}
     */
    pickMetadataIndex(scope, currentProjPath, metadataIndexDto) {
        /** @type {?} */
        const sessionId = gsp.cache.get('sessionId');
        /** @type {?} */
        const headers = new HttpHeaders({ 'Content-Type': 'application/json' }).append('SessionId', sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/pickmd?scopeType=' + scope + '&currentPath=' + currentProjPath;
        return this.http.post(url, metadataIndexDto, {
            headers
        });
    }
    // 获取元数据包中所包含的元数据信息
    /**
     * @param {?} packagePath
     * @param {?} packageName
     * @return {?}
     */
    GetMetadataPackageInfo(packagePath, packageName) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        let packageFullName = '';
        if (packagePath) {
            packageFullName = packagePath.replace(/\\/g, '/') + '/' + packageName;
        }
        else {
            packageFullName = packageName;
        }
        /** @type {?} */
        const encPackageFullName = encodeURIComponent(packageFullName);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PACKAGE_BASE_PATH + '/packagePath?packagePath=' + encPackageFullName;
        return this.http.get(url, { headers });
    }
    // 获取本地包
    /**
     * @param {?} path
     * @return {?}
     */
    GetLocalPackages(path) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PACKAGE_BASE_PATH + '/local?path=' + path;
        return this.http.get(url, { headers });
    }
    /**
     * @param {?} scope
     * @param {?=} metadataType
     * @param {?=} searchText
     * @param {?=} page
     * @param {?=} selectedPackageSource
     * @param {?=} currentProjPath
     * @return {?}
     */
    GetMetadataInfosByScope(scope, metadataType = null, searchText = null, page = null, selectedPackageSource = null, currentProjPath = '') {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        let url = this.SERVER_IP + this.METADATA_BASE_PATH + '/scope/' + scope + '?path=' + currentProjPath;
        if (page) {
            url = url + '&page=true&pageIndex=' + page.pageIndex + '&pageSize=' + page.pageSize;
        }
        else {
            url = url + '&page=false';
        }
        if (metadataType) {
            url = url + '&metadataTypeList=' + metadataType;
        }
        if (searchText) {
            url = url + '&text=' + searchText;
        }
        if ((scope === 2 || scope === 3) && selectedPackageSource) {
            url = url + '&selectedPackageSource=' + selectedPackageSource;
        }
        return this.http.get(url, { headers });
    }
    /**
     * @param {?} scope
     * @param {?=} metadataType
     * @param {?=} searchText
     * @param {?=} page
     * @param {?=} selectedPackageSource
     * @param {?=} currentProjPath
     * @return {?}
     */
    GetMetadataInfosByScope4Java(scope, metadataType = null, searchText = null, page = null, selectedPackageSource = null, currentProjPath = '') {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        let url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/mdlist?scopeType=' + scope + '&path=' + currentProjPath;
        if (page) {
            url = url + '&page=true&pageIndex=' + page.pageIndex + '&pageSize=' + page.pageSize;
        }
        else {
            url = url + '&page=false';
        }
        if (metadataType) {
            url = url + '&metadataTypeList=' + metadataType;
        }
        if (searchText) {
            url = url + '&text=' + searchText;
        }
        if ((scope === 2 || scope === 3) && selectedPackageSource) {
            url = url + '&selectedPackageSource=' + selectedPackageSource;
        }
        return this.http.get(url, { headers });
    }
    // 获取元数据以及包的信息
    // GetMetadataFromPackage(packageName: string, packagePath: string, metadataID: string): Observable<any> {
    //   const headers = new HttpHeaders().set('SessionId', this.sessionId);
    //   let fullName = '';
    //   if (packageName.endsWith('mdpkg')) {
    //     fullName = packageName;
    //   } else {
    //     fullName = packageName + '.mdpkg';
    //   }
    //   let path = '';
    //   if (packagePath) {
    //     path = packagePath;
    //   }
    //   const url = this.SERVER_IP + this.Metadata4RefDto_Base_Path + '?packageName=' + fullName +
    //     '&packagePath=' + path + '&metadataID=' + metadataID;
    //   return this.http.get(url, { headers });
    // }
    // 获取某个路径下的所有元数据包
    /**
     * @param {?} path
     * @return {?}
     */
    GetMetadataPackages(path) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.Project_Base_Path + '/children/' + path;
        return this.http.get(url, { headers });
    }
    // 获取元数据类型及后缀信息
    /**
     * @return {?}
     */
    GetMetadataTypeList() {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_CONFIG_BASE_PATH;
        return this.http.get(url, { headers });
    }
    // 获取后缀
    /**
     * @param {?} metadataType
     * @return {?}
     */
    GetPostfix(metadataType) {
        this.GetMetadataTypeList().subscribe((/**
         * @param {?} typeList
         * @return {?}
         */
        typeList => {
            typeList.forEach((/**
             * @param {?} type
             * @return {?}
             */
            type => {
                if (metadataType.toLowerCase() === type.typeCode.toLowerCase()) {
                    return type.postfix;
                }
            }));
        }));
        return new Observable();
    }
    // 更新工程引用
    /**
     * @param {?} projectPath
     * @param {?} packageHeader
     * @return {?}
     */
    UpdateMetadataProject(projectPath, packageHeader) {
        /** @type {?} */
        const content = {
            ProjectPath: projectPath,
            DepPackageName: packageHeader.name,
            DepPackageVersion: packageHeader.version.versionString,
            DepPackageLocation: packageHeader.location
        };
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH;
        return this.http.put(url, content, {
            params: {},
            headers,
            observe: 'response'
        });
    }
    // 根据业务对象ID获取app、su等信息
    /**
     * @param {?} businessObjectID
     * @param {?} sessionId
     * @return {?}
     */
    GetDevBasicBOInfo(businessObjectID, sessionId) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.DevBasinBOInfo_Base_Path + '/' + businessObjectID;
        return this.http.get(url, { headers });
    }
    /**
     * @param {?} path
     * @param {?} fileName
     * @return {?}
     */
    ValidateRepeatName(path, fileName) {
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_BASE_PATH + '/validation?path=' + path + '&fileName=' + fileName;
        /** @type {?} */
        const metadata = new Subject();
        this.http.get(url).subscribe((/**
         * @param {?} result
         * @return {?}
         */
        result => {
            // if (result) {
            metadata.next(result);
            // }
        }));
        return metadata;
    }
    /**
     * @param {?} path
     * @param {?} sessionId
     * @return {?}
     */
    SetMetadataUri(path, sessionId) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_CONFIG_BASE_PATH + '?path=' + path;
        return this.http.put(url, null, {
            params: {},
            headers,
            observe: 'response'
        });
    }
    /**
     * @return {?}
     */
    getGspProjectType() {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.GSP_PROJECT_BASE_PATH + '/type';
        return this.http.get(url, { headers });
    }
    /**
     * @param {?} path
     * @param {?} codeLanguage
     * @return {?}
     */
    updateGspProject(path, codeLanguage) {
        /** @type {?} */
        const sessionId = gsp.cache.get('sessionId');
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.GSP_PROJECT_BASE_PATH + '?path=' + path + '&codeLanguage=' + codeLanguage;
        return this.http.put(url, null, {
            params: {},
            headers,
            observe: 'response'
        });
    }
    /**
     * @return {?}
     */
    getNugetSourceList() {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.NUGET_BASE_PATH + '/sources';
        return this.http.get(url, { headers });
    }
    /**
     * @return {?}
     */
    getMavenSourceList() {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.MAVEN_JAVA_BASE_PATH + '/repolist';
        return this.http.get(url, { headers });
    }
    // 获取技术栈
    // tslint:disable-next-line:adjacent-overload-signatures
    /**
     * @param {?} projPath
     * @return {?}
     */
    getCodeLanguage(projPath) {
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/codelanguage?projPath=' + projPath;
        return this.http.get(url, { responseType: 'text' });
    }
    // 获取setting信息
    /**
     * @return {?}
     */
    getSettingInfo() {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.SETTING_BASE_PATH;
        return this.http.get(url, { headers });
    }
    // 是否开启maven
    /**
     * @return {?}
     */
    getMavenSetting() {
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_BASE_PATH + '/mavensetting';
        return this.http.get(url);
    }
    // 获取bo路径
    /**
     * @param {?} path
     * @return {?}
     */
    getBoPath(path) {
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/bopath?path=' + path;
        return this.http.get(url, { responseType: 'text' });
    }
    // 获取是否是maven引用
    /**
     * @param {?} path
     * @return {?}
     */
    getMavenStatus(path) {
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/isMaven?path=' + path;
        return this.http.get(url);
    }
    // 获取是否是解析型工具
    /**
     * @param {?} path
     * @return {?}
     */
    isInterpretation(path) {
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/isInterpretation?path=' + path;
        return this.http.get(url);
    }
    // 生成代码
    /**
     * @param {?} path
     * @return {?}
     */
    generate(path) {
        /** @type {?} */
        const url = this.JAVA_CODE_GENERATOR + '?path=' + path;
        return this.http.post(url, null);
    }
    // 获取ProcessMode
    /**
     * @param {?} path
     * @return {?}
     */
    getProcessMode(path) {
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/processMode?path=' + path;
        return this.http.get(url, { responseType: 'text' });
    }
    // 添加依赖并更新版本
    /**
     * @param {?} path
     * @param {?} metadataPackages
     * @param {?} isMaven
     * @return {?}
     */
    addDependencyAndRestore(path, metadataPackages, isMaven) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        let url;
        if (isMaven) {
            url = this.SERVER_IP + this.MAVEN_JAVA_BASE_PATH + '/adddepedency?path=' + path
                + '&metadataPackages=' + metadataPackages;
            return this.http.put(url, null, {
                params: {},
                headers,
                observe: 'response'
            });
        }
        else {
            url = this.SERVER_IP + this.NUGET_BASE_PATH + '/adddepedency?projPath=' + path
                + '&metadataPackages=' + metadataPackages;
            return this.http.post(url, null, {
                params: {},
                headers,
                observe: 'response'
            });
        }
    }
    // 在资源管理器中打开
    /**
     * @param {?} path
     * @return {?}
     */
    showInExplorer(path) {
        /** @type {?} */
        const url = this.SERVER_IP + this.FS_JAVA_BASE_PATH + '?path=' + path;
        return this.http.get(url);
    }
    /**
     * @param {?} path
     * @param {?} metadataPackages
     * @return {?}
     */
    restoreMavenPackage(path, metadataPackages) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.MAVEN_JAVA_BASE_PATH + '/adddepedency?path=' + path + '&metadataPackages=' + metadataPackages;
        return this.http.put(url, null, {
            headers,
            observe: 'response'
        });
    }
    // 获取菜单
    /**
     * @param {?} path
     * @return {?}
     */
    getMenu(path) {
        /** @type {?} */
        const url = this.SERVER_IP + this.SETTING_BASE_PATH + '/menu?path=' + path;
        return this.http.get(url);
    }
    /**
     * @param {?} projPath
     * @param {?=} fakeFlag
     * @return {?}
     */
    mavenBuildAll(projPath, fakeFlag = false) {
        if (fakeFlag) {
            return this.getMenu(projPath);
        }
        /** @type {?} */
        const url = this.JAVA_BUILD + '/compileAll?path=' + projPath;
        return this.http.post(url, null);
    }
    // 获取路径下的工程路径
    /**
     * @param {?} path
     * @return {?}
     */
    getProjPathsInPath(path) {
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/projinpath?path=' + path;
        return this.http.get(url);
    }
    /**
     * @param {?} path
     * @param {?} mdpkgName
     * @return {?}
     */
    getMdpkgNameExistInBo(path, mdpkgName) {
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_JAVA_BASE_PATH + '/mdpkgNameExist?path=' + path + '&mdpkgName=' + mdpkgName;
        return this.http.get(url, { responseType: 'text' });
    }
    // 是否存在工程文件，废弃，使用getProjPath代替
    /**
     * @param {?} path
     * @return {?}
     */
    IsProjExist(path) {
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PROJECT_BASE_PATH + '/validation?path=' + path;
        /** @type {?} */
        const metadata = new Subject();
        this.http.get(url).subscribe((/**
         * @param {?} result
         * @return {?}
         */
        result => {
            metadata.next(result);
        }));
        return metadata;
    }
    // 获取Nuget包下所有的元数据包，nuget已废弃
    /**
     * @param {?} path
     * @return {?}
     */
    GetMetadataPackagesUnderNugetDir(path) {
        /** @type {?} */
        const headers = new HttpHeaders().set('SessionId', this.sessionId);
        /** @type {?} */
        const url = this.SERVER_IP + this.METADATA_PACKAGE_BASE_PATH + '?path=' + path;
        return this.http.get(url, { headers });
    }
}
GSPMetadataService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
GSPMetadataService.ctorParameters = () => [
    { type: HttpClient },
    { type: String, decorators: [{ type: Inject, args: ['url',] }] }
];
if (false) {
    /** @type {?} */
    GSPMetadataService.prototype.selectPath;
    /** @type {?} */
    GSPMetadataService.prototype.selectType;
    /** @type {?} */
    GSPMetadataService.prototype.sessionId;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.SERVER_IP;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.METADATA_BASE_PATH;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.GSP_PROJECT_BASE_PATH;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.Project_Base_Path;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.DBO_BASE_PATH;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.METADATA_PROJECT_BASE_PATH;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.METADATA_PACKAGE_BASE_PATH;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.METADATA_CONFIG_BASE_PATH;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.DevBasinBOInfo_Base_Path;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.NUGET_BASE_PATH;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.SETTING_BASE_PATH;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.MAVEN_JAVA_BASE_PATH;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.FS_JAVA_BASE_PATH;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.JAVA_CODE_GENERATOR;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.JAVA_BUILD;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.METADATA_JAVA_BASE_PATH;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    GSPMetadataService.prototype.url;
}
