// @gsp-lcm/metadata-selector/esm2015/metadata/metadatadto.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadatadto.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class MetadataDto {
    // tslint:disable-next-line:max-line-length
    /**
     * @param {?} id
     * @param {?} nameSpace
     * @param {?} code
     * @param {?} name
     * @param {?} fileName
     * @param {?} type
     * @param {?} bizobjectID
     * @param {?} relativePath
     * @param {?} extendProperty
     * @param {?} content
     * @param {?} extendable
     */
    constructor(id, nameSpace, code, name, fileName, type, bizobjectID, relativePath, extendProperty, content, extendable) {
        this.id = id;
        this.nameSpace = nameSpace;
        this.code = code;
        this.name = name;
        this.fileName = fileName;
        this.type = type;
        this.bizobjectID = bizobjectID;
        this.relativePath = relativePath;
        this.extendProperty = extendProperty;
        this.content = content;
        this.extendable = extendable;
    }
}
if (false) {
    /** @type {?} */
    MetadataDto.prototype.id;
    /** @type {?} */
    MetadataDto.prototype.nameSpace;
    /** @type {?} */
    MetadataDto.prototype.code;
    /** @type {?} */
    MetadataDto.prototype.name;
    /** @type {?} */
    MetadataDto.prototype.fileName;
    /** @type {?} */
    MetadataDto.prototype.type;
    /** @type {?} */
    MetadataDto.prototype.bizobjectID;
    /** @type {?} */
    MetadataDto.prototype.relativePath;
    /** @type {?} */
    MetadataDto.prototype.extendProperty;
    /** @type {?} */
    MetadataDto.prototype.content;
    /** @type {?} */
    MetadataDto.prototype.extendable;
}
