// @gsp-svc/file-viewer/esm5/lib/gsp-viewer/gsp-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild } from '@angular/core';
import { StorageType } from '../file-list-mobile/entity/storagetype';
import { FileListService } from '../service/file-list.service';
import { UnifileInfo } from './entity/unifileinfo';
import { LoadingService } from '@farris/ui-loading';
import { MessagerService } from '@farris/ui-messager';
import { GspFileViewType } from '../file-list-mobile/entity/gspfileviewtype';
import { ImageViewerComponent } from '../image-viewer/image-viewer/image-viewer.component';
import { DocViewerComponent } from '../doc-viewer/doc-viewer/doc-viewer.component';
import { OfdViewerComponent } from '../ofd-viewer/ofd-viewer/ofd-viewer.component';
import { PdfViewerComponent } from '../pdf-viewer/pdf-viewer/pdf-viewer.component';
import { PdfViewOptions } from '../pdf-viewer/entitires/pdfviewoptions';
import { UrlFileInfo } from './entity/urlfileinfo';
import { CostomFileInfo } from './entity/customfileinfo';
var GspViewerComponent = /** @class */ (function () {
    function GspViewerComponent(listService, msgService, loadService) {
        this.listService = listService;
        this.msgService = msgService;
        this.loadService = loadService;
        this.viewOptions = new PdfViewOptions;
        this.isImage = true;
        this.isOfd = true;
        this.isPdf = true;
        //srcList: string[] = [];
        this.isDocShow = false;
        this.isOfdShow = false;
        this.isPdfShow = false;
    }
    /**
     * @return {?}
     */
    GspViewerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.viewFile(this.filePath, this.metadataId, this.rootId, this.fileType, this.fileName, this.serviceUnit, this.params);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    GspViewerComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.viewFile(this.filePath, this.metadataId, this.rootId, this.fileType, this.fileName, this.serviceUnit, this.params);
    };
    /**
     * @param {?} filePath
     * @param {?} metadataId
     * @param {?} rootId
     * @param {?} fileType
     * @param {?} fileName
     * @param {?} suName
     * @param {?} params
     * @return {?}
     */
    GspViewerComponent.prototype.viewFile = /**
     * @param {?} filePath
     * @param {?} metadataId
     * @param {?} rootId
     * @param {?} fileType
     * @param {?} fileName
     * @param {?} suName
     * @param {?} params
     * @return {?}
     */
    function (filePath, metadataId, rootId, fileType, fileName, suName, params) {
        var _this = this;
        if (fileType == StorageType.UniFile) {
            /** @type {?} */
            var info = new UnifileInfo;
            info.metadataId = metadataId;
            info.rootId = rootId;
            filePath = JSON.stringify(info);
        }
        if (fileType == StorageType.Url) {
            /** @type {?} */
            var info = new UrlFileInfo;
            info.fileUrl = filePath;
            info.fileName = fileName;
            filePath = JSON.stringify(info);
        }
        if (fileType == StorageType.Customer) {
            /** @type {?} */
            var info = new CostomFileInfo;
            info.suName = suName;
            info.param = params;
            filePath = JSON.stringify(info);
        }
        /** @type {?} */
        var loading = this.loadService.show({
            message: '拼命的请求中'
        });
        this.listService.getViewFile(filePath, fileType).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            loading.close();
            if (res.error != null) {
                _this.msgService.warning('获取预览文件失败' + res.error.Message);
            }
            else {
                _this.file = res;
                _this.fileChange();
            }
        }));
    };
    /**
     * @return {?}
     */
    GspViewerComponent.prototype.fileChange = /**
     * @return {?}
     */
    function () {
        //this.file=src;
        this.isImage = this.file.fileType == GspFileViewType.Image || this.file.fileType == 'Image';
        this.isOfd = this.file.fileType == GspFileViewType.Ofd || this.file.fileType == 'Ofd';
        this.isPdf = this.file.fileType == GspFileViewType.Pdf || this.file.fileType == 'Pdf';
        if (this.isImage) {
            this.isDocShow = false;
            this.isOfdShow = false; //this.imageChange(this.file.fileUrl);
            return;
        }
        else if (this.isOfd) {
            //this.file.fileUrl="/platform/runtime/dfs/images/file.jpg";
            if (this.isOfdShow)
                this.ofdChange(this.file.filePath);
            else {
                this.isDocShow = false;
                this.isOfdShow = true;
                this.isPdfShow = false;
            }
        }
        else if (this.isPdf) {
            if (this.isPdfShow)
                this.pdfChange(this.file.filePath);
            else {
                this.isDocShow = false;
                this.isOfdShow = false;
                this.isPdfShow = true;
            }
        }
        else {
            if (this.isDocShow)
                this.docChange(this.file.filePath);
            else {
                this.isDocShow = true;
                this.isOfdShow = false;
                this.isPdfShow = false;
            }
        }
    };
    /**
     * @param {?} url
     * @return {?}
     */
    GspViewerComponent.prototype.imageChange = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.imageUrl = url;
        this.imageViewer.imageChange(url);
    };
    /**
     * @param {?} filePath
     * @return {?}
     */
    GspViewerComponent.prototype.docChange = /**
     * @param {?} filePath
     * @return {?}
     */
    function (filePath) {
        this.docPath = filePath;
        this.docViewer.docChange(filePath);
    };
    /**
     * @param {?} filePath
     * @return {?}
     */
    GspViewerComponent.prototype.pdfChange = /**
     * @param {?} filePath
     * @return {?}
     */
    function (filePath) {
        //this.docPath = filePath;
        this.pdfViewer.pdfChange(filePath);
    };
    /**
     * @param {?} filePath
     * @return {?}
     */
    GspViewerComponent.prototype.ofdChange = /**
     * @param {?} filePath
     * @return {?}
     */
    function (filePath) {
        this.ofdViewer.fileChange(filePath);
    };
    GspViewerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'gsp-viewer',
                    template: "<div class=\"f-page\">\n    <div *ngIf='isImage&&!isOfd&&!isPdf'>\n        <app-image-viewer [imageUrl]=\"file.fileUrl\"></app-image-viewer>\n    </div>\n    <div *ngIf='!isImage&&!isOfd&&!isPdf'>\n        <app-doc-viewer [filePath]=\"file.filePath\"></app-doc-viewer>\n    </div>\n    <div *ngIf='!isImage&&isOfd&&!isPdf'>\n        <app-ofd-viewer [filePath]=\"file.filePath\"></app-ofd-viewer>\n    </div>\n    <div *ngIf='!isImage&&!isOfd&&isPdf' class=\"f-utils-fill\" style=\"height: 100%\">\n        <app-pdf-viewer [pdfUrl]=\"file.filePath\" [viewOptions]=\"viewOptions\"></app-pdf-viewer>\n    </div>\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    GspViewerComponent.ctorParameters = function () { return [
        { type: FileListService },
        { type: MessagerService },
        { type: LoadingService }
    ]; };
    GspViewerComponent.propDecorators = {
        filePath: [{ type: Input }],
        metadataId: [{ type: Input }],
        rootId: [{ type: Input }],
        fileName: [{ type: Input }],
        fileType: [{ type: Input }],
        viewOptions: [{ type: Input }],
        serviceUnit: [{ type: Input }],
        params: [{ type: Input }],
        imageViewer: [{ type: ViewChild, args: [ImageViewerComponent,] }],
        docViewer: [{ type: ViewChild, args: [DocViewerComponent,] }],
        ofdViewer: [{ type: ViewChild, args: [OfdViewerComponent,] }],
        pdfViewer: [{ type: ViewChild, args: [PdfViewerComponent,] }]
    };
    return GspViewerComponent;
}());
export { GspViewerComponent };
if (false) {
    /** @type {?} */
    GspViewerComponent.prototype.filePath;
    /** @type {?} */
    GspViewerComponent.prototype.metadataId;
    /** @type {?} */
    GspViewerComponent.prototype.rootId;
    /** @type {?} */
    GspViewerComponent.prototype.fileName;
    /** @type {?} */
    GspViewerComponent.prototype.fileType;
    /** @type {?} */
    GspViewerComponent.prototype.viewOptions;
    /** @type {?} */
    GspViewerComponent.prototype.serviceUnit;
    /** @type {?} */
    GspViewerComponent.prototype.params;
    /** @type {?} */
    GspViewerComponent.prototype.file;
    /** @type {?} */
    GspViewerComponent.prototype.isImage;
    /** @type {?} */
    GspViewerComponent.prototype.isOfd;
    /** @type {?} */
    GspViewerComponent.prototype.isPdf;
    /** @type {?} */
    GspViewerComponent.prototype.imageUrl;
    /** @type {?} */
    GspViewerComponent.prototype.docPath;
    /** @type {?} */
    GspViewerComponent.prototype.isDocShow;
    /** @type {?} */
    GspViewerComponent.prototype.isOfdShow;
    /** @type {?} */
    GspViewerComponent.prototype.isPdfShow;
    /** @type {?} */
    GspViewerComponent.prototype.imageViewer;
    /** @type {?} */
    GspViewerComponent.prototype.docViewer;
    /** @type {?} */
    GspViewerComponent.prototype.ofdViewer;
    /** @type {?} */
    GspViewerComponent.prototype.pdfViewer;
    /**
     * @type {?}
     * @private
     */
    GspViewerComponent.prototype.listService;
    /**
     * @type {?}
     * @private
     */
    GspViewerComponent.prototype.msgService;
    /**
     * @type {?}
     * @private
     */
    GspViewerComponent.prototype.loadService;
}
