// @gsp-sys/rtf-ui/esm2015/lib/common/rtf-ui-webapi.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class RtfUIWebapiService {
    constructor() { }
    /**
     * 菜单api地址
     * @return {?}
     */
    get functionsApi() {
        return '/api/runtime/sys/v1.0/functions/allfuncs';
    }
    /**
     * 业务对象api地址
     * @return {?}
     */
    get bizObjectApi() {
        return '/api/dev/main/v1.0/business-objects';
    }
    /**
     * 分层获取菜单数据  api
     * @return {?}
     */
    get layerFunctionsApi() {
        return '/api/runtime/sys/v1.0/functions/layerFuncs';
    }
    /**
     * Http头信息
     * @return {?}
     */
    get httpHeader() {
        /** @type {?} */
        const options = {
            headers: {
                'Content-Type': 'application/json',
            },
            responseType: 'json'
        };
        return options;
    }
}
RtfUIWebapiService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RtfUIWebapiService.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    RtfUIWebapiService.prototype.sessionId;
}
