// @edp-aif/runtime-api/esm5/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of runtime-api
 */
/**
 * runtime-api 实体类
 */
export { BefContentVariable, CreationMappingResult, CreationRuleExecutorArgs, CreationRuleExecutorAssociations, CreationRuleExecutorResult, MappingAssociation, MappingResult, ConditionResult, FormBillCreationProcessExecutorResult, RpcCreationMappingResult, RpcCreationRuleExecutorArgs, RpcCreationRuleExecutorResult, RpcFireBizEventArgs, RpcMappingResult, RpcVoChangeDetail, RpcWritebackMappingResult, RpcWritebackRuleExecutorArgs, RpcWritebackRuleExecutorResult } from "./lib/entity";
/**
 * 手工生单模块
 */
export { AifManualCreationType } from "./lib/aif-manual-creation/form-args/AifManualCreationType";
export { AifRuntimeModuleConfig } from "./lib/aif-manual-creation/config/aif-runtime-module.config";
export { AifRuntimeModule } from "./lib/aif-manual-creation/aif-runtime.module";
export { AifDataService } from "./lib/aif-manual-creation/service/aif-data.service";
export { AifSourceUiService } from "./lib/aif-manual-creation/service/aif-source-ui.service";
export { AifTargetUiService } from "./lib/aif-manual-creation/service/aif-target-ui.service";
export { SourceDataSelectorComponent } from "./lib/aif-manual-creation/components/source-data-selector/source-data-selector.component";
export { PushSelectorComponent } from "./lib/aif-manual-creation/components/push-selector/push-selector.component";
export { ReferSelectorComponent } from "./lib/aif-manual-creation/components/refer-selector/refer-selector.component";
export {} from "./lib/aif-manual-creation/base";
export { FormCreateArgs, FormCreateResult, FormCreationRuleExecutorArgs, FormEntityData, FormRestArgs, FormRestResult, FormUpdateArgs, FormUpdateResult, FormGetCreationRuleEntityExArgs, FormGetCreationRuleEntityExResult } from "./lib/aif-manual-creation/form-args";
export { AIF_UISTATE_KEY_OPEN_TYPE, AIF_UISTATE_KEY_CREATION_RULE_RESULT } from "./lib/aif-manual-creation/config/aif-form-constants";
