// @farris/ui-filter-editor/esm5/lib/components/filter-code.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ElementRef } from '@angular/core';
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @Company: Inspur
 * @Version: v0.0.1
 * @Date: 2019-04-09 11:03:49
 * @LastEditTime: 2019-08-15 17:22:59
 */
import { Component, Input, HostBinding } from '@angular/core';
var FilterCodeComponent = /** @class */ (function () {
    function FilterCodeComponent(el) {
        this.el = el;
        this.cls = 'd-flex flex-column flex-fill';
        this.conditions = [];
    }
    /**
     * @return {?}
     */
    FilterCodeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.el.nativeElement.style.overflow = 'auto';
    };
    FilterCodeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'filter-code',
                    template: "\n    <div style=\"overflow:auto\" [style.height.px]=\"height - 38\">\n        <pre class=\"flex-fill\">\n            <code>{{ conditions | json }}</code>\n        </pre>\n    </div>\n    ",
                    styles: ["\n        pre {\n            color: #d8d9d9;\n            background: #424242;\n            padding-left: 20px;\n            padding-top: 20px;\n            margin-bottom: 0;\n            font-size: 16px;\n        }\n        "]
                }] }
    ];
    /** @nocollapse */
    FilterCodeComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    FilterCodeComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        height: [{ type: Input }],
        conditions: [{ type: Input }]
    };
    return FilterCodeComponent;
}());
export { FilterCodeComponent };
if (false) {
    /** @type {?} */
    FilterCodeComponent.prototype.cls;
    /** @type {?} */
    FilterCodeComponent.prototype.height;
    /** @type {?} */
    FilterCodeComponent.prototype.conditions;
    /**
     * @type {?}
     * @private
     */
    FilterCodeComponent.prototype.el;
}
