// @farris/ui-filter-editor/esm5/lib/components/filter-footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output, HostBinding } from '@angular/core';
import { FilterGridComponent } from './filter-grid.component';
var FilterFooterComponent = /** @class */ (function () {
    function FilterFooterComponent(grid) {
        this.grid = grid;
        this.cls = 'f-filter-footer';
        this.addFilter = new EventEmitter();
        this.clearFilter = new EventEmitter();
        this.moveTop = new EventEmitter();
        this.movePrev = new EventEmitter();
        this.moveNext = new EventEmitter();
        this.moveBottom = new EventEmitter();
    }
    /**
     * @return {?}
     */
    FilterFooterComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    FilterFooterComponent.prototype.onAddFilter = /**
     * @return {?}
     */
    function () {
        this.addFilter.emit();
    };
    /**
     * @return {?}
     */
    FilterFooterComponent.prototype.onClear = /**
     * @return {?}
     */
    function () {
        this.clearFilter.emit();
    };
    /**
     * @return {?}
     */
    FilterFooterComponent.prototype.onMoveTop = /**
     * @return {?}
     */
    function () {
        this.moveTop.emit();
    };
    /**
     * @return {?}
     */
    FilterFooterComponent.prototype.onMovePrev = /**
     * @return {?}
     */
    function () {
        this.movePrev.emit();
    };
    /**
     * @return {?}
     */
    FilterFooterComponent.prototype.onMoveNext = /**
     * @return {?}
     */
    function () {
        this.moveNext.emit();
    };
    /**
     * @return {?}
     */
    FilterFooterComponent.prototype.onMoveBottom = /**
     * @return {?}
     */
    function () {
        this.moveBottom.emit();
    };
    /**
     * @return {?}
     */
    FilterFooterComponent.prototype.canUse = /**
     * @return {?}
     */
    function () {
        if (this.grid.currentCondition) {
            return this.grid.conditions.length > 1;
        }
        return false;
    };
    /**
     * @return {?}
     */
    FilterFooterComponent.prototype.canMoveUp = /**
     * @return {?}
     */
    function () {
        if (this.canUse()) {
            return this.grid.currentCondition.index > 0;
        }
        return false;
    };
    /**
     * @return {?}
     */
    FilterFooterComponent.prototype.canMoveDown = /**
     * @return {?}
     */
    function () {
        if (this.canUse()) {
            return this.grid.currentCondition.index < this.grid.conditions.length - 1;
        }
        return false;
    };
    FilterFooterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'filter-footer',
                    template: "\n    <div class=\"footer-container px-2\" style=\"height: 36px;\">\n        <button type=\"button\" (click)=\"onAddFilter()\"\n            class=\"k-button k-button-icontext k-flat \"><span class=\"f-icon f-icon-add\"></span> {{ 'filterEditor.addWhere' | locale }}</button>\n        <button type=\"button\" (click)=\"onClear()\" [disabled]=\"!grid.conditions.length\"\n            class=\"k-button k-button-icontext k-flat \"><span class=\"f-icon f-icon-close\"></span> {{ 'filterEditor.clear' | locale }}</button>\n        <button type=\"button\" [disabled]=\"!canMoveUp()\" (click)=\"onMoveTop()\"\n            class=\"k-button k-button-icontext k-flat \"><span class=\"f-icon f-icon-arrow-end-up\"></span>\n            {{'filterEditor.moveTop' | locale }}</button>\n        <button type=\"button\" [disabled]=\"!canMoveUp()\" (click)=\"onMovePrev()\"\n            class=\"k-button k-button-icontext k-flat \"><span class=\"f-icon f-icon-arrow-60-up\"></span>\n            {{'filterEditor.moveUp' | locale }}</button>\n        <button type=\"button\" [disabled]=\"!canMoveDown()\" (click)=\"onMoveNext()\"\n            class=\"k-button k-button-icontext k-flat \"><span class=\"f-icon f-icon-arrow-60-down\"></span>\n            {{'filterEditor.moveDown' | locale }}</button>\n        <button type=\"button\" [disabled]=\"!canMoveDown()\" (click)=\"onMoveBottom()\"\n            class=\"k-button k-button-icontext k-flat\"><span class=\"f-icon f-icon-arrow-end-down\"></span>\n            {{'filterEditor.moveBottom' | locale }}</button>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    FilterFooterComponent.ctorParameters = function () { return [
        { type: FilterGridComponent }
    ]; };
    FilterFooterComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        addFilter: [{ type: Output }],
        clearFilter: [{ type: Output }],
        moveTop: [{ type: Output }],
        movePrev: [{ type: Output }],
        moveNext: [{ type: Output }],
        moveBottom: [{ type: Output }]
    };
    return FilterFooterComponent;
}());
export { FilterFooterComponent };
if (false) {
    /** @type {?} */
    FilterFooterComponent.prototype.cls;
    /** @type {?} */
    FilterFooterComponent.prototype.addFilter;
    /** @type {?} */
    FilterFooterComponent.prototype.clearFilter;
    /** @type {?} */
    FilterFooterComponent.prototype.moveTop;
    /** @type {?} */
    FilterFooterComponent.prototype.movePrev;
    /** @type {?} */
    FilterFooterComponent.prototype.moveNext;
    /** @type {?} */
    FilterFooterComponent.prototype.moveBottom;
    /** @type {?} */
    FilterFooterComponent.prototype.grid;
}
