// @farris/ui-filter-editor/esm5/lib/components/filter-sql.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-05-23 10:27:07
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-08-15 17:22:46
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Input, HostBinding } from '@angular/core';
import { FilterService } from './../services/filter.service';
var FilterSqlCodeComponent = /** @class */ (function () {
    function FilterSqlCodeComponent(filterService) {
        this.filterService = filterService;
        this.cls = 'd-flex flex-column flex-fill';
        this.conditions = [];
        this.sql = '';
    }
    /**
     * @return {?}
     */
    FilterSqlCodeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.sql = this.filterService.buildSqlWhere(this.conditions);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    FilterSqlCodeComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.conditions && !changes.conditions.isFirstChange()) {
            this.sql = this.filterService.buildSqlWhere(this.conditions);
        }
    };
    FilterSqlCodeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'filter-sql',
                    template: "\n    <pre class=\"flex-fill\" [style.height.px]=\"height - 38\">\n<code>{{ sql  }}</code>\n    </pre>\n    ",
                    styles: ["\n        pre {\n            color: #d8d9d9;\n            background: #424242;\n            padding-left: 20px;\n            padding-top: 20px;\n            margin-bottom: 0;\n            font-size: 16px;\n        }\n        "]
                }] }
    ];
    /** @nocollapse */
    FilterSqlCodeComponent.ctorParameters = function () { return [
        { type: FilterService }
    ]; };
    FilterSqlCodeComponent.propDecorators = {
        height: [{ type: Input }],
        cls: [{ type: HostBinding, args: ['class',] }],
        conditions: [{ type: Input }]
    };
    return FilterSqlCodeComponent;
}());
export { FilterSqlCodeComponent };
if (false) {
    /** @type {?} */
    FilterSqlCodeComponent.prototype.height;
    /** @type {?} */
    FilterSqlCodeComponent.prototype.cls;
    /** @type {?} */
    FilterSqlCodeComponent.prototype.conditions;
    /** @type {?} */
    FilterSqlCodeComponent.prototype.sql;
    /**
     * @type {?}
     * @private
     */
    FilterSqlCodeComponent.prototype.filterService;
}
