// @farris/mobile-ui/lib/collapse/index.js
import { defineComponent, ref, watch, createVNode, resolveComponent, mergeProps, isVNode } from 'vue';

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
function __rest(s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
}

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}

var Collapse = defineComponent({
  name: 'fm-collapse',
  props: {
    isInner: Boolean,
    collapse: {
      type: Boolean,
      "default": false
    }
  },
  setup: function setup(props, _ref) {
    var attrs = _ref.attrs,
        slots = _ref.slots;
    var activeItem = ref(props.collapse ? [] : [1]);
    watch(function () {
      return props.collapse;
    }, function (value) {
      activeItem.value = value ? [] : [1];
    });

    var onChange = attrs.onChange,
        collapseAttrs = __rest(attrs, ["onChange"]);

    var itemProps = Object.assign({
      name: 1
    }, collapseAttrs);
    return function () {
      var _slot;

      return createVNode(resolveComponent("van-collapse"), {
        "modelValue": activeItem.value,
        "onUpdate:modelValue": function onUpdateModelValue($event) {
          return activeItem.value = $event;
        },
        "onChange": true,
        "class": "fm-cmp-collapse-style"
      }, _isSlot(_slot = createVNode(resolveComponent("van-collapse-item"), mergeProps(itemProps, {
        "class": props.isInner ? 'fm-tmpl-collapse-sub-header' : ''
      }), _extends({}, slots))) ? _slot : {
        "default": function _default() {
          return [_slot];
        }
      });
    };
  }
});

Collapse.install = function (app) {
  app.component(Collapse.name, Collapse);
};

export default Collapse;
export { Collapse };
