// @farris/ui-datagrid/esm5/lib/components/body/datagrid-group-footer-styler.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Renderer2, Injector, Input } from '@angular/core';
import { DatagridComponent } from '../../datagrid.component';
var GroupFooterRowStylerDirective = /** @class */ (function () {
    function GroupFooterRowStylerDirective(el, render, inject, dg) {
        this.el = el;
        this.render = render;
        this.inject = inject;
        this.dg = dg;
    }
    /**
     * @return {?}
     */
    GroupFooterRowStylerDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    GroupFooterRowStylerDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.buildCustomGroupFooterRowStyle(this.rowData);
    };
    /**
     * @private
     * @param {?} rowData
     * @return {?}
     */
    GroupFooterRowStylerDirective.prototype.buildCustomGroupFooterRowStyle = /**
     * @private
     * @param {?} rowData
     * @return {?}
     */
    function (rowData) {
        var _this = this;
        if (this.dg.groupFooterStyler) {
            /** @type {?} */
            var cs_1 = this.dg.dgs.getCustomStyle(this.dg.groupFooterStyler, [rowData]);
            if (cs_1) {
                this.dg.renderCustomCls(cs_1.cls, this.el.nativeElement);
                if (cs_1.style) {
                    Object.keys(cs_1.style).forEach((/**
                     * @param {?} n
                     * @return {?}
                     */
                    function (n) {
                        _this.el.nativeElement.style[n] = cs_1.style[n];
                    }));
                }
            }
        }
    };
    GroupFooterRowStylerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[group-footer-row-styler]',
                },] }
    ];
    /** @nocollapse */
    GroupFooterRowStylerDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: Injector },
        { type: DatagridComponent }
    ]; };
    GroupFooterRowStylerDirective.propDecorators = {
        column: [{ type: Input }],
        rowData: [{ type: Input, args: ['group-footer-row-styler',] }],
        rowIndex: [{ type: Input }]
    };
    return GroupFooterRowStylerDirective;
}());
export { GroupFooterRowStylerDirective };
if (false) {
    /** @type {?} */
    GroupFooterRowStylerDirective.prototype.column;
    /** @type {?} */
    GroupFooterRowStylerDirective.prototype.rowData;
    /** @type {?} */
    GroupFooterRowStylerDirective.prototype.rowIndex;
    /**
     * @type {?}
     * @private
     */
    GroupFooterRowStylerDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    GroupFooterRowStylerDirective.prototype.render;
    /**
     * @type {?}
     * @private
     */
    GroupFooterRowStylerDirective.prototype.inject;
    /** @type {?} */
    GroupFooterRowStylerDirective.prototype.dg;
}
