// @farris/ui-datagrid/esm5/lib/components/body/datagrid-grouprow.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Injector, Renderer2, ElementRef } from '@angular/core';
import { CUSTOM_GROUP_ROW_CLS } from '../../types/constant';
import { DatagridComponent } from './../../datagrid.component';
var DatagridGroupRowDirective = /** @class */ (function () {
    function DatagridGroupRowDirective(injector, render, el, dg) {
        this.injector = injector;
        this.render = render;
        this.el = el;
        this.dg = dg;
    }
    /**
     * @return {?}
     */
    DatagridGroupRowDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.dg && _this.dg.groupStyler) {
                /** @type {?} */
                var tr = _this.el.nativeElement;
                /** @type {?} */
                var _cls = Array.from(tr.classList).find((/**
                 * @param {?} n
                 * @return {?}
                 */
                function (n) { return n.indexOf(CUSTOM_GROUP_ROW_CLS) > -1; }));
                if (_cls) {
                    tr.classList.remove(_cls);
                }
                /** @type {?} */
                var styler = _this.dg.groupStyler(_this.groupRow);
                if (!styler) {
                    return;
                }
                if (!tr.id) {
                    tr.id = 'group-tr_' + _this.groupRow.id;
                }
                /** @type {?} */
                var str = CUSTOM_GROUP_ROW_CLS + _this.dg.idService.encrypt(JSON.stringify(styler));
                _this.render.addClass(tr, str);
                _this.dg.renderCustomStyle2(styler, tr, str);
                // this.dg.renderCustomStyle(styler, tr);
            }
        }));
    };
    DatagridGroupRowDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[group-row]',
                },] }
    ];
    /** @nocollapse */
    DatagridGroupRowDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: Renderer2 },
        { type: ElementRef },
        { type: DatagridComponent }
    ]; };
    DatagridGroupRowDirective.propDecorators = {
        groupRow: [{ type: Input, args: ['group-row',] }]
    };
    return DatagridGroupRowDirective;
}());
export { DatagridGroupRowDirective };
if (false) {
    /** @type {?} */
    DatagridGroupRowDirective.prototype.groupRow;
    /**
     * @type {?}
     * @private
     */
    DatagridGroupRowDirective.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    DatagridGroupRowDirective.prototype.render;
    /**
     * @type {?}
     * @private
     */
    DatagridGroupRowDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    DatagridGroupRowDirective.prototype.dg;
}
