// @farris/ui-datagrid/esm5/lib/components/body/datagrid-row-dblclick.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Injector, NgZone, Renderer2, HostListener } from '@angular/core';
import { of } from 'rxjs';
import { DatagridComponent } from '../../datagrid.component';
import { Utils } from '../../utils/utils';
var DatagridRowDblClickDirective = /** @class */ (function () {
    function DatagridRowDblClickDirective(el, injector, ngZone, render, dg) {
        this.el = el;
        this.injector = injector;
        this.ngZone = ngZone;
        this.render = render;
        this.dg = dg;
    }
    /**
     * @return {?}
     */
    DatagridRowDblClickDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    DatagridRowDblClickDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DatagridRowDblClickDirective.prototype.dblclickRowEvent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        /** @type {?} */
        var tr = event.target.closest('tr');
        if (!tr) {
            return;
        }
        if (tr.id == null || tr.id == undefined) {
            return;
        }
        // const rowid = tr.id.replace(TR_PREFIX, '').replace('_fixedright', '').replace('_fixedleft', '').replace('');
        /** @type {?} */
        var rowid = Utils.getRowIdByTr(tr);
        if (tr.className.indexOf('f-datagrid-group-row') > -1) {
            /** @type {?} */
            var groupRowid = rowid.replace('group-tr_', '');
            var _a = this.dg.getGroupRow(groupRowid), groupData = _a.data, groupIndex = _a.index;
            if (groupData && !isNaN(groupIndex)) {
                if (groupData.expanded) {
                    this.dg.collapseGroupRow(groupIndex);
                }
                else {
                    this.dg.expandGroupRow(groupIndex);
                }
            }
            return;
        }
        /** @type {?} */
        var rowIndex = tr.getAttribute('index');
        /** @type {?} */
        var rowData = this.dg.dfs.findRow(rowid);
        if (!rowData) {
            return;
        }
        if (this.dg.disableRow) {
            if (this.dg.disableRow(rowData.data, rowData.index)) {
                event.stopPropagation();
                return false;
            }
        }
        /** @type {?} */
        var handler = (/**
         * @return {?}
         */
        function () {
            if (!_this.dg.dfs.isRowSelected(rowid)) {
                /** @type {?} */
                var r = _this.dg.beforeSelect({ rowIndex: rowIndex, rowData: rowData.data, gridInstance: _this.dg });
                /** @type {?} */
                var r$ = of(true);
                if (r && r.subscribe) {
                    r$ = r;
                }
                r$.subscribe((/**
                 * @param {?} t
                 * @return {?}
                 */
                function (t) {
                    if (t) {
                        _this.dg.selectRow(rowid, true, false);
                        _this.dg.dblclickRow.emit({ rowData: rowData, id: rowid, index: rowIndex, grid: _this.dg, dblclick: true });
                    }
                }));
            }
            else {
                _this.dg.dblclickRow.emit({ rowData: rowData, id: rowid, index: rowIndex, grid: _this.dg, dblclick: true });
            }
        });
        if (this.dg.editable) {
            if (this.dg.editMode === 'row' && this.dg.clickToEdit) {
                this.dg.endRowEdit();
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    _this.dg.editRow(rowid);
                }), 200);
            }
            else {
                this.dg.isSingleClick = false;
                // this.dg.dblclickRow.emit({ rowData, id: rowid, index: rowIndex, grid: this.dg, dblclick: true });
                handler();
            }
        }
        else {
            // this.dg.dblclickRow.emit({ rowData, id: rowid, index: rowIndex, grid: this.dg,  dblclick: true });
            handler();
        }
        event.stopPropagation();
        event.preventDefault();
    };
    DatagridRowDblClickDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[row-dblclick]',
                },] }
    ];
    /** @nocollapse */
    DatagridRowDblClickDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Injector },
        { type: NgZone },
        { type: Renderer2 },
        { type: DatagridComponent }
    ]; };
    DatagridRowDblClickDirective.propDecorators = {
        dblclickRowEvent: [{ type: HostListener, args: ['dblclick', ['$event'],] }]
    };
    return DatagridRowDblClickDirective;
}());
export { DatagridRowDblClickDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DatagridRowDblClickDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    DatagridRowDblClickDirective.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    DatagridRowDblClickDirective.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    DatagridRowDblClickDirective.prototype.render;
    /** @type {?} */
    DatagridRowDblClickDirective.prototype.dg;
}
