// @edp-aif/common-api/esm2015/lib/billkit/BillData.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from '../BpmnModelJsonHelper';
import { KitBill } from './KitBill';
import { BillDataTable } from './BillDataTable';
export class BillData extends KitBill {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.mainTable = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "mainTable", this.mainTable);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.mainTable = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "mainTable", BillDataTable);
        this.ApplyRelationship();
    }
    /*
         * 将数据集、主表、子表、行、字段之间的隶属关系应用到各个对象上。
         * */
    /**
     * @return {?}
     */
    ApplyRelationship() {
        this.Apply2Table(this, null, this.mainTable);
        return this;
    }
    /**
     * @private
     * @param {?} attachedBillData
     * @param {?} attachedRow
     * @param {?} table
     * @return {?}
     */
    Apply2Table(attachedBillData, attachedRow, table) {
        table.attachedBillData = attachedBillData;
        table.attachedRow = attachedRow;
        for (let row of table.GetRowList()) {
            this.Apply2Row(table, row);
        }
    }
    /**
     * @private
     * @param {?} attachedTable
     * @param {?} row
     * @return {?}
     */
    Apply2Row(attachedTable, row) {
        row.attachedTable = attachedTable;
        for (let field of row.GetFieldList()) {
            this.Apply2Field(row, field);
        }
        for (let childTable of row.GetChildTableList()) {
            this.Apply2Table(attachedTable.attachedBillData, row, childTable);
        }
    }
    /**
     * @private
     * @param {?} attachedRow
     * @param {?} field
     * @return {?}
     */
    Apply2Field(attachedRow, field) {
        field.attachedRow = attachedRow;
        for (let subField of field.GetSubFieldList()) {
            this.Apply2SubField(field, subField);
        }
    }
    /**
     * @private
     * @param {?} attachedField
     * @param {?} field
     * @return {?}
     */
    Apply2SubField(attachedField, field) {
        field.attachedField = field;
        for (let subField of field.GetSubFieldList()) {
            this.Apply2SubField(field, subField);
        }
    }
}
if (false) {
    /** @type {?} */
    BillData.prototype.mainTable;
}
