// @edp-aif/common-api/esm2015/lib/billkit/BillSchemaTables.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class BillSchemaTables {
    constructor() {
        this._map = {};
        this._list = new Array();
    }
    /**
     * @return {?}
     */
    get map() {
        return this._map;
    }
    /**
     * @return {?}
     */
    get list() {
        return this._list;
    }
    /**
     * @return {?}
     */
    GetMap() {
        return this._map;
    }
    /**
     * @return {?}
     */
    GetList() {
        return this._list;
    }
    /**
     * @param {?} table
     * @return {?}
     */
    Add(table) {
        this._map[table.tableName] = table;
        this._list.push(table);
    }
    /**
     * @param {?} table
     * @return {?}
     */
    Remove(table) {
        /** @type {?} */
        const bak = this._map[table.tableName];
        delete this._map[table.tableName];
        for (let i = 0; i < this._list.length; i++) {
            if (this._list[i].tableName == bak.tableName) {
                this._list.splice(i, 1);
                return true;
            }
        }
        return false;
    }
    /**
     * @return {?}
     */
    Clear() {
        this._map = {};
        this._list.length = 0;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    BillSchemaTables.prototype._map;
    /**
     * @type {?}
     * @private
     */
    BillSchemaTables.prototype._list;
}
