// @edp-aif/common-api/esm2015/lib/billkit/BillDataRows.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class BillDataRows {
    constructor() {
        this._map = {};
        this._list = new Array();
    }
    /**
     * @return {?}
     */
    get map() {
        return this._map;
    }
    /**
     * @return {?}
     */
    get list() {
        return this._list;
    }
    /**
     * @param {?} row
     * @return {?}
     */
    Add(row) {
        this._map[row.dataId] = row;
        this._list.push(row);
    }
    /**
     * @param {?} row
     * @return {?}
     */
    Remove(row) {
        /** @type {?} */
        const bak = this._map[row.dataId];
        delete this._map[row.dataId];
        for (let i = 0; i < this._list.length; i++) {
            if (this._list[i].dataId == bak.dataId) {
                this._list.splice(i, 1);
                return true;
            }
        }
        return false;
    }
    /**
     * @return {?}
     */
    Clear() {
        this._map = {};
        this._list.length = 0;
    }
    /**
     * @return {?}
     */
    GetMap() {
        return this._map;
    }
    /**
     * @return {?}
     */
    GetList() {
        return this._list;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    BillDataRows.prototype._map;
    /**
     * @type {?}
     * @private
     */
    BillDataRows.prototype._list;
}
