// @edp-aif/common-api/esm2015/lib/billkit/BillDataFields.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class BillDataFields {
    constructor() {
        this._map = {};
        this._list = new Array();
    }
    /**
     * @return {?}
     */
    get map() {
        return this._map;
    }
    /**
     * @return {?}
     */
    get list() {
        return this._list;
    }
    /**
     * @param {?} field
     * @return {?}
     */
    Add(field) {
        this._map[field.fieldName] = field;
        this._list.push(field);
    }
    /**
     * @param {?} field
     * @return {?}
     */
    Remove(field) {
        /** @type {?} */
        const bak = this._map[field.fieldName];
        delete this._map[field.fieldName];
        for (let i = 0; i < this._list.length; i++) {
            if (this._list[i].fieldName == bak.fieldName) {
                this._list.splice(i, 1);
                return true;
            }
        }
        return false;
    }
    /**
     * @return {?}
     */
    Clear() {
        this._map = {};
        this._list.length = 0;
    }
    /**
     * @return {?}
     */
    GetMap() {
        return this._map;
    }
    /**
     * @return {?}
     */
    GetList() {
        return this._list;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    BillDataFields.prototype._map;
    /**
     * @type {?}
     * @private
     */
    BillDataFields.prototype._list;
}
