// @edp-aif/common-api/esm2015/lib/billkit/BillDataTable.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from '../BpmnModelJsonHelper';
import { BillDataRow } from './BillDataRow';
import { BillDataRows } from './BillDataRows';
import { KitTable } from './KitTable';
export class BillDataTable extends KitTable {
    constructor() {
        super(...arguments);
        this.rows = new BillDataRows();
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.attachedBillData = null;
        this.attachedRow = null;
        this.rows.Clear();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        /** @type {?} */
        let dataRows = [];
        for (let value of this.rows.GetList()) {
            dataRows.push(value.ConvertToJsonObject());
        }
        BpmnModelJsonHelper.SetAttributeValue(obj, "rows", dataRows);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        /** @type {?} */
        let dataRows = BpmnModelJsonHelper.GetAttributeValue(obj, "rows");
        for (let item of dataRows) {
            /** @type {?} */
            let value = new BillDataRow();
            value.LoadFromJsonObject(item);
            this.rows.Add(value);
        }
    }
    /**
     * @param {?} row
     * @return {?}
     */
    AddRow(row) {
        this.rows.Add(row);
    }
    /**
     * @return {?}
     */
    GetRowMap() {
        return this.rows.GetMap();
    }
    /**
     * @return {?}
     */
    GetRowList() {
        return this.rows.GetList();
    }
}
if (false) {
    /** @type {?} */
    BillDataTable.prototype.rows;
    /** @type {?} */
    BillDataTable.prototype.attachedBillData;
    /** @type {?} */
    BillDataTable.prototype.attachedRow;
}
