// @edp-aif/common-api/esm2015/lib/billkit/BillSchemaTable.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from '../BpmnModelJsonHelper';
import { BillSchemaFields } from './BillSchemaFields';
import { BillSchemaField } from './BillSchemaField';
import { KitTable } from './KitTable';
import { BillSchemaTables } from './BillSchemaTables';
export class BillSchemaTable extends KitTable {
    constructor() {
        super(...arguments);
        this.fields = new BillSchemaFields();
        this.childTables = new BillSchemaTables();
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.caption = null;
        this.attachedBillSchema = null;
        this.attachedTable = null;
        this.fields.Clear();
        this.childTables.Clear();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "caption", this.caption);
        /** @type {?} */
        let fields = [];
        for (let value of this.fields.GetList()) {
            fields.push(value.ConvertToJsonObject());
        }
        BpmnModelJsonHelper.SetAttributeValue(obj, "fields", fields);
        if (this.childTables.GetList().length > 0) {
            /** @type {?} */
            let childTables = [];
            for (let value of this.childTables.GetList()) {
                childTables.push(value.ConvertToJsonObject());
            }
            BpmnModelJsonHelper.SetAttributeValue(obj, "childTables", childTables);
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.caption = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "caption");
        /** @type {?} */
        let fields = BpmnModelJsonHelper.GetAttributeValue(obj, "fields");
        for (let item of fields) {
            /** @type {?} */
            let value = new BillSchemaField();
            value.LoadFromJsonObject(item);
            this.fields.Add(value);
        }
        /** @type {?} */
        let childTables = BpmnModelJsonHelper.GetAttributeValue(obj, "childTables");
        if (childTables != null) {
            for (let item of childTables) {
                /** @type {?} */
                let value = new BillSchemaTable();
                value.LoadFromJsonObject(item);
                this.childTables.Add(value);
            }
        }
    }
    /**
     * @param {?} field
     * @return {?}
     */
    AddField(field) {
        this.fields.Add(field);
    }
    /**
     * @return {?}
     */
    getFieldMap() {
        return this.fields.GetMap();
    }
    /**
     * @return {?}
     */
    getFieldList() {
        return this.fields.GetList();
    }
    /**
     * @param {?} table
     * @return {?}
     */
    AddChildTable(table) {
        this.childTables.Add(table);
    }
    /**
     * @return {?}
     */
    GetChildTableMap() {
        return this.childTables.GetMap();
    }
    /**
     * @return {?}
     */
    GetChildTableList() {
        return this.childTables.GetList();
    }
}
if (false) {
    /** @type {?} */
    BillSchemaTable.prototype.caption;
    /** @type {?} */
    BillSchemaTable.prototype.fields;
    /** @type {?} */
    BillSchemaTable.prototype.childTables;
    /** @type {?} */
    BillSchemaTable.prototype.attachedBillSchema;
    /** @type {?} */
    BillSchemaTable.prototype.attachedTable;
}
