// @edp-aif/common-api/esm2015/lib/billkit/BillSchema.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from '../BpmnModelJsonHelper';
import { KitBill } from './KitBill';
import { BillSchemaTable } from './BillSchemaTable';
export class BillSchema extends KitBill {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.mainTable = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "mainTable", this.mainTable);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.mainTable = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "mainTable", BillSchemaTable);
        this.ApplyRelationship();
    }
    /*
         * 将架构、主表、子表、列、字段之间的隶属关系应用到各个对象上。
         * */
    /**
     * @return {?}
     */
    ApplyRelationship() {
        this.Apply2Table(this, null, this.mainTable);
        return this;
    }
    /**
     * @private
     * @param {?} attachedBillSchema
     * @param {?} attachedTable
     * @param {?} table
     * @return {?}
     */
    Apply2Table(attachedBillSchema, attachedTable, table) {
        table.attachedBillSchema = attachedBillSchema;
        table.attachedTable = attachedTable;
        for (let field of table.getFieldList()) {
            this.Apply2Field(table, field);
        }
        for (let childTable of table.GetChildTableList()) {
            this.Apply2Table(table.attachedBillSchema, table, childTable);
        }
    }
    /**
     * @private
     * @param {?} attachedTable
     * @param {?} field
     * @return {?}
     */
    Apply2Field(attachedTable, field) {
        field.attachedTable = attachedTable;
        for (let subField of field.GetSubFieldList()) {
            this.Apply2SubField(field, subField);
        }
    }
    /**
     * @private
     * @param {?} attachedField
     * @param {?} field
     * @return {?}
     */
    Apply2SubField(attachedField, field) {
        field.attachedField = attachedField;
        for (let subField of field.GetSubFieldList()) {
            this.Apply2SubField(field, subField);
        }
    }
}
if (false) {
    /** @type {?} */
    BillSchema.prototype.mainTable;
}
