// @edp-aif/common-api/esm2015/lib/billkit/KitBill.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from '../BpmnModelJsonHelper';
import { JsonConvertableImpl } from '../JsonConvertableImpl';
/**
 * @abstract
 */
export class KitBill extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.schemaCategoryId = null;
        this.schemaId = null;
        this.schemaCaption = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaCategoryId", this.schemaCategoryId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaId", this.schemaId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaCaption", this.schemaCaption);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.schemaCategoryId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaCategoryId");
        this.schemaId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaId");
        this.schemaCaption = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaCaption");
    }
}
if (false) {
    /** @type {?} */
    KitBill.prototype.schemaCategoryId;
    /** @type {?} */
    KitBill.prototype.schemaId;
    /** @type {?} */
    KitBill.prototype.schemaCaption;
}
