// @edp-aif/common-api/esm2015/lib/billkit/KitFieldInstance.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { KitField } from './KitField';
import { DataTypeStoreImpl } from './DataTypeStoreImpl';
import { ExtensionMethod } from '../ExtensionMethod';
/**
 * @abstract
 */
export class KitFieldInstance extends KitField {
    /**
     * @param {?=} fieldName
     * @param {?=} fieldValue
     */
    constructor(fieldName, fieldValue) {
        super();
        this.fieldStore = new DataTypeStoreImpl();
        this.fieldName = fieldName;
        this.SetFieldValue(fieldValue);
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.fieldStore = new DataTypeStoreImpl();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        if ("base.string" != this.fieldStore.converterName) {
            ExtensionMethod.setAttributeValue(obj, "conv", this.fieldStore.converterName);
        }
        if (this.fieldStore.textValue != null) {
            ExtensionMethod.setAttributeValue(obj, "value", this.fieldStore.textValue);
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        /** @type {?} */
        let conv = ExtensionMethod.getAttributeValue4String(obj, 'conv');
        if (conv == null) {
            conv = "base.string";
        }
        this.fieldStore.converterName = conv;
        /** @type {?} */
        let text = ExtensionMethod.getAttributeValue4String(obj, "value");
        if (text != null) {
            this.fieldStore.textValue = text;
        }
    }
    /**
     * @return {?}
     */
    GetFieldValue() {
        return this.fieldStore.dataValue;
    }
    /**
     * @param {?} fieldValue
     * @return {?}
     */
    SetFieldValue(fieldValue) {
        this.fieldStore.dataValue = fieldValue;
    }
    /**
     * @return {?}
     */
    GetTextValue() {
        return this.fieldStore.textValue;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    KitFieldInstance.prototype.fieldStore;
}
