// @ecp-caf/caf-common/esm5/lib/services/httpservice/http.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { SessionService } from '../sessionservice/session.service';
import { CommonUtil } from '../../utilities/common.utilities';
var HttpService = /** @class */ (function () {
    function HttpService(http, sessionService) {
        this.http = http;
        this.sessionService = sessionService;
    }
    /**
     * @param {?} url
     * @return {?}
     */
    HttpService.prototype.get = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return this.request('Get', url, this.setHeader({}));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    HttpService.prototype.post = /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        return this.request('Post', url, this.setHeader({ body: body }));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    HttpService.prototype.put = /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        return this.request('Put', url, this.setHeader({ body: body }));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    HttpService.prototype.delete = /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        return this.request('Delete', url, this.setHeader({ body: body }));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    HttpService.prototype.patch = /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        return this.request('Patch', url, this.setHeader({ body: body }));
    };
    /** 默认用当前sessionid */
    /**
     * 默认用当前sessionid
     * @param {?} options
     * @param {?=} bizContextId
     * @return {?}
     */
    HttpService.prototype.setHeader = /**
     * 默认用当前sessionid
     * @param {?} options
     * @param {?=} bizContextId
     * @return {?}
     */
    function (options, bizContextId) {
        options.headers = {
            'Content-Type': 'application/json'
        };
        /** @type {?} */
        var commonVariables = CommonUtil.getCommonVariables();
        if (commonVariables) {
            /** @type {?} */
            var commonVariableKey = 'X-CAF-Runtime-CommonVariable';
            options.headers[commonVariableKey] = commonVariables;
        }
        if (bizContextId) {
            /** @type {?} */
            var bizContextKey = 'X-CAF-Runtime-Context';
            options.headers[bizContextKey] = bizContextId;
        }
        return options;
    };
    /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    HttpService.prototype.request = /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    function (method, url, options) {
        if (options.body) {
            if (typeof options.body !== 'string') {
                options.body = JSON.stringify(options.body);
            }
        }
        return this.http.request(method, url, options)
            .pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            return data;
        })));
    };
    /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    HttpService.prototype.request$ = /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    function (method, url, options) {
        return this.http.request(method, url, options);
    };
    HttpService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    HttpService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: SessionService }
    ]; };
    return HttpService;
}());
export { HttpService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    HttpService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    HttpService.prototype.sessionService;
}
