// @farris/ui-datagrid/esm5/lib/types/constant.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-06 07:43:07
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-11-11 10:48:38
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { InjectionToken } from '@angular/core';
/** @type {?} */
export var GRID_EDITORS = new InjectionToken('Farris DataGrid Editors.');
/** @type {?} */
export var GRID_VALIDATORS = new InjectionToken('Farris DataGrid Validators.');
/** @type {?} */
export var GRID_FILTER_ROW = new InjectionToken('Farris DataGrid Filter-Row Component.');
/** @type {?} */
export var GRID_SETTINGS_PANEL = new InjectionToken('Farris DataGrid Setting Panel Component.');
/** @type {?} */
export var GRID_SETTINGS_SERVICE = new InjectionToken('Farris DataGrid Setting Service.');
/** @type {?} */
export var GRID_SETTINGS_HTTP = new InjectionToken(' Datagrid Settings HttpClient.');
/** @type {?} */
export var SCROLL_Y_ACTION = 'psScrollY';
/** @type {?} */
export var SCROLL_X_ACTION = 'psScrollX';
/** @type {?} */
export var SCROLL_UP_ACTION = 'psScrollUp';
/** @type {?} */
export var SCROLL_DOWN_ACTION = 'psScrollDown';
/** @type {?} */
export var SCROLL_LEFT_ACTION = 'psScrollLeft';
/** @type {?} */
export var SCROLL_RIGHT_ACTION = 'psScrollRight';
/** @type {?} */
export var SCROLL_Y_REACH_END_ACTION = 'psYReachEnd';
/** @type {?} */
export var SCROLL_Y_REACH_START_ACTION = 'psYReachStart';
/** @type {?} */
export var SCROLL_X_REACH_END_ACTION = 'psXReachEnd';
/** @type {?} */
export var SCROLL_X_REACH_START_ACTION = 'psXReachStart';
/** @type {?} */
export var DATAGRID_SETTING_FIELDNAME = '_DATAGRID_CONTROL_PANEL_';
/**
 * 左侧固定列阴影样式名称
 * @type {?}
 */
export var FIXED_LEFT_SHADOW_CLS = 'f-datagrid-fixed-left-shadow';
/** @type {?} */
export var FIXED_RIGHT_SHADOW_CLS = 'f-datagrid-fixed-right-shadow';
/** @type {?} */
export var ROW_HOVER_CLS = 'f-datagrid-row-hover';
/** @type {?} */
export var ROW_SELECTED_CLS = 'f-datagrid-row-selected';
/** @type {?} */
export var CELL_SELECTED_CLS = 'f-datagrid-cell-selected';
/** @type {?} */
export var CELL_EDITING_CLS = 'f-datagrid-cell-editing';
/**
 * 行高尺寸定义
 * @type {?}
 */
export var SIZE_TYPE = {
    sm: {
        row: 29,
        // 行高
        header: 35,
        // 表头行高
        pager: 40 // 分页条高度
    },
    md: {
        row: 36,
        header: 40,
        pager: 44
    },
    lg: {
        row: 46,
        header: 40,
        pager: 44
    },
    xl: {
        row: 60,
        header: 40,
        pager: 44
    },
};
/**
 * 允许编辑时，醒目显示可编辑单元格的样式
 * @type {?}
 */
export var ALLOW_EDIT_CELL_HIGHLIGHT_CLS = {
    datagrid: 'highlight-allowedit-cell',
    textarea: 'cell-textarea',
    textbox: 'cell-textbox',
    checkbox: 'cell-checkbox',
    datepicker: 'cell-datepicker',
    'input-group': 'cell-input-group',
    select: 'cell-combolist',
    lookup: 'cell-lookup',
    numberbox: 'cell-numberbox',
    combolist: 'cell-combolist',
    'combo-lookup': 'cell-combolist',
    switch: 'cell-switch',
    timepicker: 'cell-timepicker',
    'language-textbox': 'cell-language-textbox',
    'enum-editor': 'cell-enum-editor'
};
/** @type {?} */
export var TR_PREFIX = 'row_';
/** @type {?} */
export var TR_ID_PREFIX = 'row_';
/** @type {?} */
export var CUSTOM_CELL_CLS = 'customcellcls_';
/** @type {?} */
export var CUSTOM_GROUP_ROW_CLS = 'customgrouprowcls_';
/** @type {?} */
export var CUSTOM_ROW_CLS = 'customrowcls_';
/**
 * 折行高度cls 前缀 wrap_row_height_
 * @type {?}
 */
export var WORDWRAP_ROW_HEIGHT_CLS = 'wrap_row_height_';
