// @gsp-bef/gsp-vo-metadata/esm5/lib/view-model/src/Action/ParaEnum/VMParameterType.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var VMParameterType = {
    /**
     * 字符型
     */
    String: 'String',
    /**
     * 布尔型
     */
    Boolean: 'Boolean',
    /**
     * 整数
     */
    Int32: 'Int32',
    /**
     * 浮点数
     */
    Decimal: 'Decimal',
    /**
     * 双浮点
     */
    Double: 'Double',
    /**
     * 时间
     */
    DateTime: 'DateTime',
    /**
     * 对象类型
     */
    Object: 'Object',
    /**
     * 自定义
     */
    Custom: 'Custom',
};
export { VMParameterType };
/** @type {?} */
export var VMParameterTypes = [
    { index: 0, text: '字符串', value: VMParameterType.String, assembly: 'mscorlib.dll', className: 'System.String', javaClassName: 'java.lang.String' },
    { index: 1, text: '布尔型', value: VMParameterType.Boolean, assembly: 'mscorlib.dll', className: 'System.Boolean', javaClassName: 'java.lang.Boolean' },
    { index: 2, text: '整数', value: VMParameterType.Int32, assembly: 'mscorlib.dll', className: 'System.Int32', javaClassName: 'java.lang.Integer' },
    { index: 3, text: '浮点数', value: VMParameterType.Decimal, assembly: 'mscorlib.dll', className: 'System.Single', javaClassName: 'java.math.BigDecimal' },
    { index: 4, text: '双浮点数', value: VMParameterType.Double, assembly: 'mscorlib.dll', className: 'System.Double', javaClassName: 'java.lang.Double' },
    { index: 5, text: '时间类型', value: VMParameterType.DateTime, assembly: 'mscorlib.dll', className: 'System.DateTime', javaClassName: 'java.util.Date' },
    { index: 6, text: '对象类型', value: VMParameterType.Object, assembly: 'mscorlib.dll', className: 'System.Object', javaClassName: 'java.lang.Object' },
    { index: 7, text: '自定义类型', value: VMParameterType.Custom, assembly: '', className: '', javaClassName: '' },
];
