// @farris/component-querycondition/esm5/lib/entity/conditionvalue/checkbox-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/checkbox-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
// CheckBoxValue初始化，与'app-solution-checkbox'组件初始化是分离的
// CheckBoxValue 要求value是布尔类型，组件要求value是true时，值应为true字符串
var 
// CheckBoxValue初始化，与'app-solution-checkbox'组件初始化是分离的
// CheckBoxValue 要求value是布尔类型，组件要求value是true时，值应为true字符串
CheckBoxValue = /** @class */ (function () {
    // CheckBoxValue 初始化，传入的值未做任何修改
    function CheckBoxValue(value) {
        if (value === void 0) { value = []; }
        this.value = value;
    }
    /**
     * @return {?}
     */
    CheckBoxValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = [];
    };
    // 如果组件的返回值是字符串，转化为布尔值
    // 如果组件的返回值是字符串，转化为布尔值
    /**
     * @param {?} value
     * @return {?}
     */
    CheckBoxValue.prototype.setOriginalValue = 
    // 如果组件的返回值是字符串，转化为布尔值
    /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value.length) {
            this.value = value.map((/**
             * @param {?} v
             * @return {?}
             */
            function (v) {
                if (typeof v == 'string') {
                    return true;
                }
                else {
                    return v;
                }
            }));
        }
        else {
            this.value = value;
        }
    };
    // 把CheckBoxValue的true布尔值，转化为组件可识别的字符串
    // 把CheckBoxValue的true布尔值，转化为组件可识别的字符串
    /**
     * @return {?}
     */
    CheckBoxValue.prototype.getOriginalValue = 
    // 把CheckBoxValue的true布尔值，转化为组件可识别的字符串
    /**
     * @return {?}
     */
    function () {
        if (this.value.length) {
            return this.value.map((/**
             * @param {?} v
             * @return {?}
             */
            function (v) {
                if (v === true) {
                    return "true";
                }
                else {
                    return v;
                }
            }));
        }
        return this.value;
    };
    /**
     * @return {?}
     */
    CheckBoxValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.BoolCheck;
    };
    /**
     * @return {?}
     */
    CheckBoxValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return this.value.length == 0;
    };
    return CheckBoxValue;
}());
// CheckBoxValue初始化，与'app-solution-checkbox'组件初始化是分离的
// CheckBoxValue 要求value是布尔类型，组件要求value是true时，值应为true字符串
export { CheckBoxValue };
if (false) {
    /** @type {?} */
    CheckBoxValue.prototype.value;
}
