// @farris/component-querycondition/esm5/lib/entity/conditionvalue/combolookup-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/combolookup-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
import { cloneDeep } from 'lodash-es';
var ComboLookUpValue = /** @class */ (function () {
    function ComboLookUpValue(data) {
        if (data === void 0) { data = { value: [], valueField: undefined, textValue: undefined }; }
        if (data.textValue) { //新结构，value是数组
            this.valueField = data.valueField;
            this.value = data.value;
            this.textValue = data.textValue;
        }
        else { //旧结构，value是对象
            this.valueField = data.valueField ? data.valueField : '';
            this.value = (data.value && data.value.data) ? [data.value.data] : [];
            this.textValue = 'help-text-value-null'; //旧结构无textValue属性
        }
    }
    /**
     * @return {?}
     */
    ComboLookUpValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = [];
        this.valueField = undefined;
        this.textValue = undefined;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ComboLookUpValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        throw new Error("Method not implemented.");
    };
    /**
     * @return {?}
     */
    ComboLookUpValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var args = this.valueField.split('.');
        /** @type {?} */
        var valueArr = this.value.map((/**
         * @param {?} helpItem
         * @return {?}
         */
        function (helpItem) { return _this.getPropValue(helpItem, cloneDeep(args)); }));
        return valueArr && valueArr.length ? valueArr.join(',') : '';
    };
    /**
     * @return {?}
     */
    ComboLookUpValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.ComboLookUp;
    };
    /**
     * @return {?}
     */
    ComboLookUpValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.valueField;
    };
    /**
     * @param {?} helpItem
     * @param {?} args
     * @return {?}
     */
    ComboLookUpValue.prototype.getPropValue = /**
     * @param {?} helpItem
     * @param {?} args
     * @return {?}
     */
    function (helpItem, args) {
        if (args.length > 1) {
            /** @type {?} */
            var arg = args.shift();
            return helpItem[arg] ? this.getPropValue(helpItem[arg], args) : null;
        }
        return helpItem[args[0]];
    };
    /**
     * @param {?} textField
     * @return {?}
     */
    ComboLookUpValue.prototype.getTextValue = /**
     * @param {?} textField
     * @return {?}
     */
    function (textField) {
        var _this = this;
        /** @type {?} */
        var args = textField.split('.');
        /** @type {?} */
        var textArr = this.value.map((/**
         * @param {?} helpItem
         * @return {?}
         */
        function (helpItem) { return _this.getPropValue(helpItem, cloneDeep(args)); }));
        return textArr && textArr.length ? textArr.join(',') : '';
    };
    return ComboLookUpValue;
}());
export { ComboLookUpValue };
if (false) {
    /** @type {?} */
    ComboLookUpValue.prototype.valueField;
    /** @type {?} */
    ComboLookUpValue.prototype.value;
    /** @type {?} */
    ComboLookUpValue.prototype.textValue;
}
