// @farris/component-querycondition/esm5/lib/entity/conditionvalue/dropdownlistvalue.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/dropdownlistvalue.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
var DropDownListValue = /** @class */ (function () {
    function DropDownListValue(data) {
        if (data === void 0) { data = { value: [], key: undefined }; }
        this.value = []; //旧结构{Type: '', Content: {value: string, name: string}} 新结构Array[{value: string, name: string}]
        if (data.value && data.value.length > 0) { //已有初始值[]
            if (data.hasOwnProperty('key')) { //新结构
                this.value = data.value;
                this.key = data.key;
            }
            else { //旧结构
                this.value = data.value['Content'] && data.value['Content'].value ? [data.value['Content']] : [];
                this.key = data.value['Content'] && data.value['Content'].value ? data.value['Content'].value : undefined;
            }
        }
        else {
            this.clearValue();
        }
    }
    /**
     * @return {?}
     */
    DropDownListValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = [];
        this.key = undefined;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    DropDownListValue.prototype.setOriginalValue = /**
     * @param {?} data
     * @return {?}
     */
    function (data) { };
    /**
     * @return {?}
     */
    DropDownListValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.key;
    };
    /**
     * @return {?}
     */
    DropDownListValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.DropDownList;
    };
    /**
     * @return {?}
     */
    DropDownListValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.key;
    };
    return DropDownListValue;
}());
export { DropDownListValue };
if (false) {
    /** @type {?} */
    DropDownListValue.prototype.value;
    /** @type {?} */
    DropDownListValue.prototype.key;
}
