// @farris/component-querycondition/esm5/lib/entity/conditionvalue/inputgroupvalue.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/inputgroupvalue.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { cloneDeep } from 'lodash-es';
var InputGroupValue = /** @class */ (function () {
    function InputGroupValue(data) {
        if (data === void 0) { data = { value: [], textValue: undefined, textField: undefined, isInputText: undefined }; }
        this.value = data.value;
        // this.textValue = data.textValue ? data.textValue.trim() : '';
        this.textValue = data.textValue ? data.textValue : '';
        this.textField = data.textField;
        this.isInputText = data.isInputText;
    }
    /**
     * @return {?}
     */
    InputGroupValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.InputGroup;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    InputGroupValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        throw new Error("Method not implemented.");
    };
    /**
     * @return {?}
     */
    InputGroupValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var args = this.textField ? this.textField.split('.') : [];
        /** @type {?} */
        var textArr = this.value.map((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return _this.getPropValue(item, cloneDeep(args)); }));
        return textArr && textArr.length ? textArr.join(',') : '';
    };
    /**
     * @return {?}
     */
    InputGroupValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = [];
        this.textValue = undefined;
        this.textField = undefined;
        this.isInputText = undefined;
    };
    /**
     * @return {?}
     */
    InputGroupValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.textValue;
    };
    /**
     * @param {?} targetField
     * @return {?}
     */
    InputGroupValue.prototype.getTextValue = /**
     * @param {?} targetField
     * @return {?}
     */
    function (targetField) {
        var _this = this;
        /** @type {?} */
        var args = targetField.split('.');
        /** @type {?} */
        var textArr = this.value.map((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return _this.getPropValue(item, cloneDeep(args)); }));
        return textArr && textArr.length ? textArr.join(',') : '';
    };
    /**
     * @param {?} item
     * @param {?} args
     * @return {?}
     */
    InputGroupValue.prototype.getPropValue = /**
     * @param {?} item
     * @param {?} args
     * @return {?}
     */
    function (item, args) {
        if (args.length > 1) {
            /** @type {?} */
            var arg = args.shift();
            return item[arg] ? this.getPropValue(item[arg], args) : '';
        }
        return item[args[0]];
    };
    return InputGroupValue;
}());
export { InputGroupValue };
if (false) {
    /** @type {?} */
    InputGroupValue.prototype.value;
    /** @type {?} */
    InputGroupValue.prototype.textValue;
    /** @type {?} */
    InputGroupValue.prototype.textField;
    /** @type {?} */
    InputGroupValue.prototype.isInputText;
}
