// @farris/component-querycondition/esm5/lib/entity/conditionvalue/numberrangevaue.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/numberrangevaue.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
var NumberRangeValue = /** @class */ (function () {
    function NumberRangeValue(data) {
        if (data === void 0) { data = { startValue: null, endValue: null }; }
        this.startValue = data.startValue == null ? null : parseFloat(data.startValue);
        this.endValue = data.endValue == null ? null : parseFloat(data.endValue);
    }
    /**
     * @return {?}
     */
    NumberRangeValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return (this.startValue == null && this.endValue == null);
    };
    /**
     * @return {?}
     */
    NumberRangeValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.startValue = undefined;
        this.endValue = undefined;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NumberRangeValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        throw new Error("Method not implemented.");
    };
    /**
     * @return {?}
     */
    NumberRangeValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return {
            begin: this.startValue,
            end: this.endValue
        };
    };
    /**
     * @return {?}
     */
    NumberRangeValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.NumberRange;
    };
    return NumberRangeValue;
}());
export { NumberRangeValue };
if (false) {
    /** @type {?} */
    NumberRangeValue.prototype.startValue;
    /** @type {?} */
    NumberRangeValue.prototype.endValue;
}
var NumberRangeValueConvertor = /** @class */ (function () {
    function NumberRangeValueConvertor() {
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    NumberRangeValueConvertor.prototype.convertToObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var numberRangeValue = (/** @type {?} */ (obj));
        /** @type {?} */
        var jobj = new Object();
        jobj["startValue"] = numberRangeValue.startValue;
        jobj["endValue"] = numberRangeValue.endValue;
        return jobj;
    };
    /**
     * @param {?} jobj
     * @return {?}
     */
    NumberRangeValueConvertor.prototype.initFromObject = /**
     * @param {?} jobj
     * @return {?}
     */
    function (jobj) {
        /** @type {?} */
        var numberRangeValue = new NumberRangeValue();
        if (jobj["startValue"]) {
            //key point！as Number并不会转化格式
            numberRangeValue.startValue = parseFloat(jobj["startValue"]);
        }
        if (jobj["endValue"]) {
            numberRangeValue.endValue = parseFloat(jobj["endValue"]);
        }
        return numberRangeValue;
    };
    return NumberRangeValueConvertor;
}());
export { NumberRangeValueConvertor };
