// @farris/component-querycondition/esm5/lib/entity/conditionvalue/radio-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/radio-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
var RadioValue = /** @class */ (function () {
    function RadioValue(value) {
        if (value === void 0) { value = undefined; }
        this.value = value;
    }
    /**
     * @return {?}
     */
    RadioValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = undefined;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    RadioValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = value;
    };
    /**
     * @return {?}
     */
    RadioValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.value;
    };
    /**
     * @return {?}
     */
    RadioValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.Radio;
    };
    /**
     * @return {?}
     */
    RadioValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.value;
    };
    return RadioValue;
}());
export { RadioValue };
if (false) {
    /** @type {?} */
    RadioValue.prototype.value;
}
