// @farris/component-querycondition/esm5/lib/entity/conditionvalue/single-date-time-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/single-date-time-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
var SingleDateTimeValue = /** @class */ (function () {
    function SingleDateTimeValue(datetime) {
        if (datetime === void 0) { datetime = ''; }
        this.datetimeValue = datetime;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    SingleDateTimeValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.datetimeValue = value.formatted;
    };
    /**
     * @return {?}
     */
    SingleDateTimeValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.datetimeValue;
    };
    /**
     * @return {?}
     */
    SingleDateTimeValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.SingleDateTime;
    };
    /**
     * @return {?}
     */
    SingleDateTimeValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.datetimeValue;
    };
    /**
     * @return {?}
     */
    SingleDateTimeValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.datetimeValue = undefined;
    };
    return SingleDateTimeValue;
}());
export { SingleDateTimeValue };
if (false) {
    /** @type {?} */
    SingleDateTimeValue.prototype.datetimeValue;
}
