// @farris/component-querycondition/esm5/lib/entity/conditionvalue/single-number-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/single-number-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
var SingleNumberValue = /** @class */ (function () {
    function SingleNumberValue(data) {
        if (data === void 0) { data = null; }
        this.numValue = data == null ? null : parseFloat(data);
    }
    /**
     * @return {?}
     */
    SingleNumberValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return this.numValue == null;
    };
    /**
     * @return {?}
     */
    SingleNumberValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.numValue = undefined;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    SingleNumberValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        throw new Error("Method not implemented.");
    };
    /**
     * @return {?}
     */
    SingleNumberValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.numValue;
    };
    /**
     * @return {?}
     */
    SingleNumberValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.SingleNumber;
    };
    return SingleNumberValue;
}());
export { SingleNumberValue };
if (false) {
    /** @type {?} */
    SingleNumberValue.prototype.numValue;
}
