// @farris/component-querycondition/esm5/lib/entity/conditionvalue/single-month-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/single-month-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
var SingleMonthValue = /** @class */ (function () {
    function SingleMonthValue(month) {
        if (month === void 0) { month = ''; }
        this.monthValue = month;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    SingleMonthValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.monthValue = value.formatted;
    };
    /**
     * @return {?}
     */
    SingleMonthValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.monthValue;
    };
    /**
     * @return {?}
     */
    SingleMonthValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.SingleMonth;
    };
    /**
     * @return {?}
     */
    SingleMonthValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.monthValue = undefined;
    };
    /**
     * @return {?}
     */
    SingleMonthValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.monthValue;
    };
    return SingleMonthValue;
}());
export { SingleMonthValue };
if (false) {
    /** @type {?} */
    SingleMonthValue.prototype.monthValue;
}
