// @farris/component-querycondition/esm5/lib/entity/conditionvalue/single-year-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/single-year-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
var SingleYearValue = /** @class */ (function () {
    function SingleYearValue(year) {
        if (year === void 0) { year = ''; }
        this.yearValue = year;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    SingleYearValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.yearValue = value.formatted;
    };
    /**
     * @return {?}
     */
    SingleYearValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.yearValue;
    };
    /**
     * @return {?}
     */
    SingleYearValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.SingleYear;
    };
    /**
     * @return {?}
     */
    SingleYearValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.yearValue = undefined;
    };
    /**
     * @return {?}
     */
    SingleYearValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.yearValue;
    };
    return SingleYearValue;
}());
export { SingleYearValue };
if (false) {
    /** @type {?} */
    SingleYearValue.prototype.yearValue;
}
