// @farris/component-querycondition/esm5/lib/entity/conditionvalue/smarthelpvalue.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/smarthelpvalue.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
import { cloneDeep } from 'lodash-es';
var SmartHelpValue = /** @class */ (function () {
    function SmartHelpValue(data) {
        if (data === void 0) { data = { value: [], valueField: undefined, textValue: undefined, isInputText: undefined }; }
        if (data.textValue) {
            this.value = data.value;
            this.valueField = data.valueField;
            this.textValue = data.textValue;
            this.isInputText = data.isInputText;
        }
        else {
            //帮助未支持多选时，没有textValue属性，所有值都是通过选择产生的
            this.value = data.value;
            this.valueField = data.valueField ? data.valueField : '';
            this.textValue = 'help-text-value-null';
            this.isInputText = false;
        }
    }
    /**
     * @return {?}
     */
    SmartHelpValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.SmartHelp;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    SmartHelpValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        throw new Error("Method not implemented.");
    };
    /**
     * @return {?}
     */
    SmartHelpValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var args = this.valueField.split('.');
        /** @type {?} */
        var valueArr = this.value.map((/**
         * @param {?} helpItem
         * @return {?}
         */
        function (helpItem) { return _this.getPropValue(helpItem, cloneDeep(args)); }));
        return valueArr && valueArr.length ? valueArr.join(',') : '';
    };
    /**
     * @return {?}
     */
    SmartHelpValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = [];
        this.valueField = undefined;
        this.textValue = undefined;
        this.isInputText = undefined;
    };
    /**
     * @return {?}
     */
    SmartHelpValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.value.length;
    };
    /**
     * @param {?} textField
     * @return {?}
     */
    SmartHelpValue.prototype.getTextValue = /**
     * @param {?} textField
     * @return {?}
     */
    function (textField) {
        var _this = this;
        /** @type {?} */
        var args = textField.split('.');
        /** @type {?} */
        var textArr = this.value.map((/**
         * @param {?} helpItem
         * @return {?}
         */
        function (helpItem) { return _this.getPropValue(helpItem, cloneDeep(args)); }));
        return textArr && textArr.length ? textArr.join(',') : '';
    };
    /**
     * @param {?} helpItem
     * @param {?} args
     * @return {?}
     */
    SmartHelpValue.prototype.getPropValue = /**
     * @param {?} helpItem
     * @param {?} args
     * @return {?}
     */
    function (helpItem, args) {
        if (args.length > 1) {
            /** @type {?} */
            var arg = args.shift();
            return helpItem[arg] ? this.getPropValue(helpItem[arg], args) : null;
        }
        return helpItem[args[0]];
    };
    return SmartHelpValue;
}());
export { SmartHelpValue };
if (false) {
    /** @type {?} */
    SmartHelpValue.prototype.value;
    /** @type {?} */
    SmartHelpValue.prototype.valueField;
    /** @type {?} */
    SmartHelpValue.prototype.textValue;
    /** @type {?} */
    SmartHelpValue.prototype.isInputText;
}
