// @farris/component-querycondition/esm5/lib/entity/conditionvalue/textvalue.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/textvalue.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
var TextValue = /** @class */ (function () {
    function TextValue(data) {
        if (data === void 0) { data = { value: '' }; }
        // this.value = data.value ? data.value.trim() : '';
        this.value = data.value ? data.value : '';
    }
    /**
     * @return {?}
     */
    TextValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = undefined;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TextValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = value;
    };
    /**
     * @return {?}
     */
    TextValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.value;
    };
    /**
     * @return {?}
     */
    TextValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.Text;
    };
    /**
     * @return {?}
     */
    TextValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.value;
    };
    return TextValue;
}());
export { TextValue };
if (false) {
    /** @type {?} */
    TextValue.prototype.value;
}
