// @farris/ui-datagrid-editors/esm5/lib/editors/datagrid-enumeditor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Renderer2, ElementRef, NgZone, Injector, ViewChild } from '@angular/core';
import { EnumEditorInputComponent } from '@farris/ui-enum-editor';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { EnumEditorDefaultOptions } from '../editor-default-options';
var DatagridEnumEditorComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridEnumEditorComponent, _super);
    function DatagridEnumEditorComponent(render, el, ngzone, injector) {
        var _this = _super.call(this, render, el, injector) || this;
        _this.ngzone = ngzone;
        _this.injector = injector;
        return _this;
    }
    /**
     * @return {?}
     */
    DatagridEnumEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.options = Object.assign({}, EnumEditorDefaultOptions, this.options);
    };
    /**
     * @return {?}
     */
    DatagridEnumEditorComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.inputElement = this.instance.inputGroupRef.textbox.nativeElement;
        _super.prototype.ngAfterViewInit.call(this);
    };
    DatagridEnumEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'datagrid-enum-editor',
                    template: "\n    <div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup farris-group-auto\">\n    <enum-editor-input #eei\n        style=\"width: 100%\"\n        [formControlName]=\"column.field\"\n        [readonly]=\"options.readonly\"\n        [editable]=\"options.editable\"\n        [disabled]=\"options.disabled\"\n        [enableClear]=\"options.enableClear\"\n        [outType]=\"options.outType\"\n        [textField]=\"options.textField\"\n        [valueField]=\"options.valueField\">\n    </enum-editor-input>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    DatagridEnumEditorComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: NgZone },
        { type: Injector }
    ]; };
    DatagridEnumEditorComponent.propDecorators = {
        instance: [{ type: ViewChild, args: ['eei',] }]
    };
    return DatagridEnumEditorComponent;
}(DatagridBaseEditorDirective));
export { DatagridEnumEditorComponent };
if (false) {
    /** @type {?} */
    DatagridEnumEditorComponent.prototype.instance;
    /**
     * @type {?}
     * @private
     */
    DatagridEnumEditorComponent.prototype.ngzone;
    /** @type {?} */
    DatagridEnumEditorComponent.prototype.injector;
}
