// @farris/ui-datagrid-editors/esm5/lib/editors/datagrid-select.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-14 11:41:00
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-10-02 17:07:03
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Renderer2, ElementRef, ViewChild, Injector } from '@angular/core';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
var DatagridSelectComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridSelectComponent, _super);
    function DatagridSelectComponent(render, el, injector) {
        var _this = _super.call(this, render, el, injector) || this;
        _this.enumData = [];
        return _this;
    }
    /**
     * @return {?}
     */
    DatagridSelectComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.inputElement = this.instance.nativeElement;
        if (this.options) {
            this.enumData = this.options.data;
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridSelectComponent.prototype.onClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        $event.stopPropagation();
        return false;
    };
    DatagridSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'grid-editor-select',
                    template: "\n    <div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup\">\n        <select #sel class=\"custom-select\" [formControlName]=\"column.field\" (click)=\"onClick($event)\" >\n            <option *ngFor=\"let col of enumData\" value=\"{{col[options.valueField]}}\">{{ col[options.textField] }}</option>\n        </select>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    DatagridSelectComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: Injector }
    ]; };
    DatagridSelectComponent.propDecorators = {
        instance: [{ type: ViewChild, args: ['sel',] }]
    };
    return DatagridSelectComponent;
}(DatagridBaseEditorDirective));
export { DatagridSelectComponent };
if (false) {
    /** @type {?} */
    DatagridSelectComponent.prototype.enumData;
    /** @type {?} */
    DatagridSelectComponent.prototype.instance;
}
