// @farris/ui-datagrid-editors/esm5/lib/editors/datagrid-switch.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-14 11:41:00
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-08-24 15:33:24
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Renderer2, ElementRef, Injector, ViewChild } from '@angular/core';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { SwitchComponent } from '@farris/ui-switch';
import { SwitchDefaultOptions } from '../editor-default-options';
var DatagridSwitchComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridSwitchComponent, _super);
    function DatagridSwitchComponent(render, el, injector) {
        return _super.call(this, render, el, injector) || this;
    }
    /**
     * @return {?}
     */
    DatagridSwitchComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.options = Object.assign({}, SwitchDefaultOptions, this.options);
        if (this.options && !this.options.valueChange) {
            this.options.valueChange = (/**
             * @param {?} val
             * @return {?}
             */
            function (val) { });
        }
        /** @type {?} */
        var align = this.column.align || 'left';
        if (align === 'left') {
            this.render.setStyle(this.ctrlBoxEl.nativeElement, 'justify-content', align);
            this.render.setStyle(this.ctrlBoxEl.nativeElement, 'padding-left', '0.75rem');
        }
        this.inputElement = this.instance.switchElRef.nativeElement;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridSwitchComponent.prototype.onValueChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var p = this.eventParams($event);
        p['checked'] = this.instance.checked;
        p.value = this.instance['getValue'](this.instance.checked);
        p['instance'] = this.instance;
        p['editor'] = this;
        this.options.valueChange(p);
    };
    DatagridSwitchComponent.decorators = [
        { type: Component, args: [{
                    selector: 'grid-editor-switch',
                    template: "\n    <div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup farris-group-auto\" style=\"line-height:100%\" #ctrlbox hidefocus=\"true\">\n        <farris-switch #switch [formControlName]=\"column.field\"\n        [square]=\"options.square\"\n        [readonly]=\"options.readonly\"\n        [disabled]=\"options.disabled\"\n        [checkedLabel]=\"options.checkedLabel\"\n        [uncheckedLabel]=\"options.uncheckedLabel\"\n        [trueValue]=\"options?.trueValue\"\n        [falseValue]=\"options?.falseValue\"\n        (valueChange)=\"onValueChange($event)\"></farris-switch>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    DatagridSwitchComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: Injector }
    ]; };
    DatagridSwitchComponent.propDecorators = {
        ctrlBoxEl: [{ type: ViewChild, args: ['ctrlbox',] }],
        instance: [{ type: ViewChild, args: ['switch',] }]
    };
    return DatagridSwitchComponent;
}(DatagridBaseEditorDirective));
export { DatagridSwitchComponent };
if (false) {
    /** @type {?} */
    DatagridSwitchComponent.prototype.ctrlBoxEl;
    /** @type {?} */
    DatagridSwitchComponent.prototype.instance;
}
