// @farris/ui-datagrid-editors/esm5/lib/editors/datagrid-numberspinner.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-14 11:41:00
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-10-22 14:25:39
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Renderer2, ElementRef, Injector, ViewChild } from '@angular/core';
import { NumberSpinnerComponent } from '@farris/ui-number-spinner';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { NumberSpinnerDefaultOptions } from '../editor-default-options';
var DatagridNumberSpinnerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridNumberSpinnerComponent, _super);
    function DatagridNumberSpinnerComponent(render, el, injector) {
        return _super.call(this, render, el, injector) || this;
    }
    /**
     * @return {?}
     */
    DatagridNumberSpinnerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.options = Object.assign({}, NumberSpinnerDefaultOptions, this.options);
        if (!this.options.textAlign) {
            this.options.textAlign = this.column.align || 'left';
        }
        this.inputElement = this.instance.input.nativeElement;
        // this.stopPropagation = false;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridNumberSpinnerComponent.prototype.onValueChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.options.valueChange) {
            this.options.valueChange(this.eventPrams($event));
        }
    };
    /**
     * @private
     * @param {?} $event
     * @return {?}
     */
    DatagridNumberSpinnerComponent.prototype.eventPrams = /**
     * @private
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var p = this.eventParams($event);
        p['instance'] = this.instance;
        p['editor'] = this;
        return p;
    };
    DatagridNumberSpinnerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'grid-editor-numberspinner',
                    template: "\n    <div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup farris-group-auto\">\n        <datagrid-tooltip [control]=\"formControl\" [tooltipPosition]=\"'top-left'\" [message]=\"errorMessage\">\n            <farris-number-spinner #num [step]=\"options.step\"\n                [max]=\"options.max\"\n                [min]=\"options.min\"\n                [canNull]=\"options.canNull\"\n                [precision]=\"options.precision\"\n                [editable]=\"options.editable\"\n                [bigNumber]=\"options.bigNumber\"\n                [showButton]=\"options.showButton\"\n                [showZero]=\"options.showZero\"\n                [groupSeparator]=\"options.groupSeparator\"\n                (valueChange)=\"onValueChange($event)\"\n                [textAlign]=\"options.textAlign\"\n                [formControlName]=\"column.field\" style=\"width: 100%\"></farris-number-spinner>\n        </datagrid-tooltip>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    DatagridNumberSpinnerComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: Injector }
    ]; };
    DatagridNumberSpinnerComponent.propDecorators = {
        instance: [{ type: ViewChild, args: ['num',] }]
    };
    return DatagridNumberSpinnerComponent;
}(DatagridBaseEditorDirective));
export { DatagridNumberSpinnerComponent };
if (false) {
    /** @type {?} */
    DatagridNumberSpinnerComponent.prototype.instance;
}
