// @farris/ui-datagrid-editors/esm5/lib/editors/datagrid-filtereditor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Renderer2, ElementRef, NgZone, Injector, ViewChild } from '@angular/core';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { FilterComponent } from '@farris/ui-filter-editor';
import { FilterEditorDefaultOptions } from '../editor-default-options';
var DatagridFilterEditorComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridFilterEditorComponent, _super);
    function DatagridFilterEditorComponent(render, el, ngzone, injector) {
        var _this = _super.call(this, render, el, injector) || this;
        _this.ngzone = ngzone;
        _this.injector = injector;
        return _this;
    }
    /**
     * @return {?}
     */
    DatagridFilterEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.options = Object.assign({}, FilterEditorDefaultOptions, this.options);
    };
    /**
     * @return {?}
     */
    DatagridFilterEditorComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.inputElement = this.instance.textbox.nativeElement;
        _super.prototype.ngAfterViewInit.call(this);
    };
    DatagridFilterEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'datagrid-filter-editor',
                    template: "\n    <div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup farris-group-auto\">\n        <filter #filterEditor\n            style=\"width: 100%\"\n            [formControlName]=\"column.field\"\n            [columns]=\"options.columns\"\n            [showCode]=\"options.showCode\"\n            [showSql]=\"options.showSql\"\n            [disabled]=\"options.disabled\"\n            [readonly]=\"options.readonly\"\n            [editable]=\"options.editable\"\n            [enableClear]=\"options.enableClear\"\n            [fieldEditor]=\"options.fieldEditor\"\n            [enableExpress]=\"options.enableExpress\"\n            [returnType]=\"options.returnType\"\n            (showExpress)=\"options?.onshowExpress($event)\"></filter>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    DatagridFilterEditorComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: NgZone },
        { type: Injector }
    ]; };
    DatagridFilterEditorComponent.propDecorators = {
        instance: [{ type: ViewChild, args: ['filterEditor',] }]
    };
    return DatagridFilterEditorComponent;
}(DatagridBaseEditorDirective));
export { DatagridFilterEditorComponent };
if (false) {
    /** @type {?} */
    DatagridFilterEditorComponent.prototype.instance;
    /**
     * @type {?}
     * @private
     */
    DatagridFilterEditorComponent.prototype.ngzone;
    /** @type {?} */
    DatagridFilterEditorComponent.prototype.injector;
}
