// @farris/ui-datagrid-editors/esm5/lib/editors/datagrid-inputgroup.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { EventEmitter, Injector } from '@angular/core';
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-14 11:41:00
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-10-14 13:04:50
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Renderer2, ElementRef, ViewChild, Output } from '@angular/core';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { InputGroupComponent } from '@farris/ui-input-group';
import { InputGroupDefaultOptions } from '../editor-default-options';
var DatagridInputGroupComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridInputGroupComponent, _super);
    function DatagridInputGroupComponent(render, el, injector) {
        var _this = _super.call(this, render, el, injector) || this;
        _this.injector = injector;
        _this.clear = new EventEmitter();
        _this.clickHandle = new EventEmitter();
        return _this;
    }
    /**
     * @return {?}
     */
    DatagridInputGroupComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.inputElement = this.instance.textbox.nativeElement;
        this.options = Object.assign({}, InputGroupDefaultOptions, this.options);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridInputGroupComponent.prototype.onClear = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.options.clear) {
            this.options.clear.call(this, this.eParams($event));
        }
        else {
            this.clear.emit();
        }
        this.instance.cd.detectChanges();
        return false;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridInputGroupComponent.prototype.onClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.options.clickHandle) {
            this.options.clickHandle.call(this, this.eParams($event));
        }
        else {
            this.clickHandle.emit();
        }
        this.instance.cd.detectChanges();
        return false;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridInputGroupComponent.prototype.onValueChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.options.valueChange) {
            this.options.valueChange.call(this, this.eParams($event));
        }
    };
    /**
     * @private
     * @param {?} $event
     * @return {?}
     */
    DatagridInputGroupComponent.prototype.eParams = /**
     * @private
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var p = this.eventParams($event);
        p['$event'] = $event;
        p['instance'] = this.instance;
        p['editor'] = this;
        p.value = $event ? $event.value : '';
        return p;
    };
    DatagridInputGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'grid-editor-inputgroup',
                    template: "\n    <div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup farris-group-auto\">\n        <input-group #ig style=\"width: 100%\"\n            (clear)=\"onClear($event)\"\n            (clickHandle)=\"onClick($event)\"\n            (valueChange)=\"onValueChange($event)\"\n            [readonly] = \"options.readonly\"\n            [editable] = \"options.editable\"\n            [enableClear] = \"options.showClear\"\n            [groupText] = \"options.groupText\"\n            [isPassword] = \"options.isPassword\"\n            [textAlign]=\"options.textAlign\"\n            [enableViewPassword]=\"options.enableViewPassword\"\n            [formControlName]=\"column.field\"\n            [maxLength]=\"options.maxLength\"\n            ></input-group>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    DatagridInputGroupComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: Injector }
    ]; };
    DatagridInputGroupComponent.propDecorators = {
        clear: [{ type: Output }],
        clickHandle: [{ type: Output }],
        instance: [{ type: ViewChild, args: ['ig',] }]
    };
    return DatagridInputGroupComponent;
}(DatagridBaseEditorDirective));
export { DatagridInputGroupComponent };
if (false) {
    /** @type {?} */
    DatagridInputGroupComponent.prototype.clear;
    /** @type {?} */
    DatagridInputGroupComponent.prototype.clickHandle;
    /** @type {?} */
    DatagridInputGroupComponent.prototype.instance;
    /** @type {?} */
    DatagridInputGroupComponent.prototype.injector;
}
