// @farris/ui-datagrid-editors/esm5/lib/editors/datagrid-languagetextbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Renderer2, ElementRef, Injector, ViewChild } from '@angular/core';
import { LanguageTextboxComponent } from '@farris/ui-language-textbox';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { LanguageTextboxDefaultOptions } from '../editor-default-options';
var DatagridLanguageTextboxComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridLanguageTextboxComponent, _super);
    function DatagridLanguageTextboxComponent(render, el, injector) {
        return _super.call(this, render, el, injector) || this;
    }
    /**
     * @return {?}
     */
    DatagridLanguageTextboxComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.inputElement = this.instance.input.textbox.nativeElement;
        this.options = Object.assign({}, LanguageTextboxDefaultOptions, this.options);
        if (this.options && !this.options.valueChange) {
            this.options.valueChange = (/**
             * @param {?} val
             * @return {?}
             */
            function (val) { });
        }
    };
    DatagridLanguageTextboxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'grid-editor-language-textbox',
                    template: "\n    <div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup farris-group-auto\" style=\"line-height:100%\">\n        <language-textbox #ins style=\"width: 100%\"\n            [languages]=\"options.languages\" [formControlName]=\"column.field\"\n            [enableClear]=\"options.enableClear\"\n            [readonly]=\"options.readonly\"\n            [currentLanguage]=\"options.currentLanguage\"\n            ></language-textbox>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    DatagridLanguageTextboxComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: Injector }
    ]; };
    DatagridLanguageTextboxComponent.propDecorators = {
        instance: [{ type: ViewChild, args: ['ins',] }]
    };
    return DatagridLanguageTextboxComponent;
}(DatagridBaseEditorDirective));
export { DatagridLanguageTextboxComponent };
if (false) {
    /** @type {?} */
    DatagridLanguageTextboxComponent.prototype.instance;
}
