// @farris/ui-datagrid-editors/esm5/lib/editors/datagrid-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-14 11:44:10
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-10-14 12:46:12
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Renderer2, ElementRef, NgZone, ViewChild, Injector } from '@angular/core';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
var DatagridCheckboxComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridCheckboxComponent, _super);
    function DatagridCheckboxComponent(render, el, ngzone, injector) {
        var _this = _super.call(this, render, el, injector) || this;
        _this.ngzone = ngzone;
        _this.injector = injector;
        return _this;
    }
    /**
     * @return {?}
     */
    DatagridCheckboxComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.inputElement = this.instance.nativeElement;
        this.checked = this.formControl.value;
        /** @type {?} */
        var align = this.column.align || 'left';
        this.render.setStyle(this.chkboxEl.nativeElement, 'justify-content', align);
        this.render.listen(this.inputElement, 'keydow.Sapce', this.onSpaceHandle);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridCheckboxComponent.prototype.onSpaceHandle = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        $event.preventDefault();
        $event.stopPropagation();
        this.checkedChanges();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DatagridCheckboxComponent.prototype.handleClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.stopPropagation();
        this.checkedChanges();
    };
    /**
     * @private
     * @return {?}
     */
    DatagridCheckboxComponent.prototype.checkedChanges = /**
     * @private
     * @return {?}
     */
    function () {
        this.checked = !this.checked;
        this.formControl.setValue(this.checked);
        this.formControl.markAsDirty();
        if (this.options && this.options.valueChange) {
            /** @type {?} */
            var p = this.eventParams(event);
            p.value = this.checked;
            p['instance'] = this.instance;
            p['editor'] = this;
            this.options.valueChange(p);
        }
    };
    DatagridCheckboxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'grid-editor-checkbox',
                    template: "<div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup\" #chkbox>\n    <div class=\"custom-control custom-checkbox f-checkradio-single\">\n        <input type=\"checkbox\" #chk class=\"custom-control-input\" [formControlName]=\"column.field\" [checked]=\"checked\">\n        <label class=\"custom-control-label\" (click)=\"handleClick($event)\"></label>\n    </div></div>\n    "
                }] }
    ];
    /** @nocollapse */
    DatagridCheckboxComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: NgZone },
        { type: Injector }
    ]; };
    DatagridCheckboxComponent.propDecorators = {
        instance: [{ type: ViewChild, args: ['chk',] }],
        chkboxEl: [{ type: ViewChild, args: ['chkbox',] }]
    };
    return DatagridCheckboxComponent;
}(DatagridBaseEditorDirective));
export { DatagridCheckboxComponent };
if (false) {
    /** @type {?} */
    DatagridCheckboxComponent.prototype.checked;
    /** @type {?} */
    DatagridCheckboxComponent.prototype.instance;
    /** @type {?} */
    DatagridCheckboxComponent.prototype.chkboxEl;
    /**
     * @type {?}
     * @private
     */
    DatagridCheckboxComponent.prototype.ngzone;
    /** @type {?} */
    DatagridCheckboxComponent.prototype.injector;
}
