// @farris/ui-datagrid-editors/esm5/lib/editors/datagrid-datepicker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Renderer2, ElementRef, NgZone, ViewChild, Injector } from '@angular/core';
import { DatagridBaseEditorDirective } from './../datagrid-base-editor.directive';
import { FarrisDatepickerComponent } from '@farris/ui-datepicker';
import { DatePickerDefaultOptions } from '../editor-default-options';
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-14 11:40:36
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-10-14 12:58:14
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
var DatagridDatepickerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridDatepickerComponent, _super);
    function DatagridDatepickerComponent(render, el, ngzone, injector) {
        var _this = _super.call(this, render, el, injector) || this;
        _this.ngzone = ngzone;
        _this.injector = injector;
        return _this;
    }
    /**
     * @return {?}
     */
    DatagridDatepickerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.options = Object.assign({}, DatePickerDefaultOptions, this.options);
    };
    /**
     * @return {?}
     */
    DatagridDatepickerComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.inputElement = this.instance.dateInput ? this.instance.dateInput.nativeElement : this.el.nativeElement.querySelector('input');
        _super.prototype.ngAfterViewInit.call(this);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridDatepickerComponent.prototype.onClear = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.options.clear) {
            this.options.clear(this.eventPrams($event));
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridDatepickerComponent.prototype.onValueChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.options.valueChange) {
            this.options.valueChange(this.eventPrams($event));
        }
        if (this.inputElement) {
            this.inputElement.focus();
        }
    };
    /**
     * @private
     * @param {?} $event
     * @return {?}
     */
    DatagridDatepickerComponent.prototype.eventPrams = /**
     * @private
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var p = this.eventParams($event);
        p['instance'] = this.instance;
        p['editor'] = this;
        return p;
    };
    DatagridDatepickerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'grid-editor-datepicker',
                    template: "\n    <div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup farris-group-auto\">\n        <farris-datepicker\n            #datepicker\n            style=\"width: 100%\"\n            [formControlName]=\"column.field\"\n            [readonly]=\"options.readonly\"\n            [editable]=\"options.editable\"\n            [locale]=\"options.locale\"\n            [dateRange]=\"options.dateRange\"\n            [showTime]=\"options.showTime\"\n            [showType]=\"options.showType\"\n            [dateFormat]=\"options.dateFormat\"\n            [returnFormat]=\"options.returnFormat\"\n            [returnType]=\"options.returnType\"\n            [useDefault]=\"options.useDefault\"\n            [placeholder]=\"options.placeholder\"\n            [maxDate]=\"options.maxDate\"\n            [minDate]=\"options.minDate\"\n            [dateRangeDatesDelimiter]=\"options.dateRangeDatesDelimiter\"\n            [showWeekNumbers]=\"options?.showWeekNumbers\"\n            [shortcuts]=\"[]\"\n            (clear)=\"onClear($event)\"\n            (valueChange)=\"onValueChange($event)\"\n            [viewType]=\"options.viewType\"\n            [showPresent]=\"options.showPresent\"\n            [hourStep]=\"options.hourStep\"\n            [minuteStep]=\"options.minuteStep\"\n            [secondStep]=\"options.secondStep\"\n            [firstDayOfWeek]=\"options.firstDayOfWeek\"\n        ></farris-datepicker>\n    </div>\n\n\n    "
                }] }
    ];
    /** @nocollapse */
    DatagridDatepickerComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: NgZone },
        { type: Injector }
    ]; };
    DatagridDatepickerComponent.propDecorators = {
        instance: [{ type: ViewChild, args: ['datepicker',] }]
    };
    return DatagridDatepickerComponent;
}(DatagridBaseEditorDirective));
export { DatagridDatepickerComponent };
if (false) {
    /** @type {?} */
    DatagridDatepickerComponent.prototype.instance;
    /**
     * @type {?}
     * @private
     */
    DatagridDatepickerComponent.prototype.ngzone;
    /** @type {?} */
    DatagridDatepickerComponent.prototype.injector;
}
