// @farris/ui-datagrid-editors/esm5/lib/editors/datagrid-timer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-14 11:41:00
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-08-24 15:33:24
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Renderer2, ElementRef, Injector, ViewChild } from '@angular/core';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { TimePickerComponent } from '@farris/ui-time-picker';
import { TimePickerDefaultOptions } from '../editor-default-options';
var DatagridTimerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridTimerComponent, _super);
    function DatagridTimerComponent(render, el, injector) {
        return _super.call(this, render, el, injector) || this;
    }
    /**
     * @return {?}
     */
    DatagridTimerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.inputElement = this.instance.timeInput.nativeElement;
        this.options = Object.assign({}, TimePickerDefaultOptions, this.options);
        if (this.options && !this.options.valueChange) {
            this.options.valueChange = (/**
             * @param {?} val
             * @return {?}
             */
            function (val) { });
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridTimerComponent.prototype.onValueChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var p = this.eventParams($event);
        p.value = this.instance.value;
        p['instance'] = this.instance;
        p['editor'] = this;
        this.options.valueChange(p);
    };
    DatagridTimerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'grid-editor-timer',
                    template: "\n    <div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup farris-group-auto\" >\n        <farris-time-picker #timepicker\n            style=\"width: 100%\"\n            [formControlName]=\"column.field\"\n            [readonly]= \"options.readonly\"\n            [editable]= \"options.editable\"\n            [hourStep] = \"options.hourStep\"\n            [minuteStep] = \"options.minuteStep\"\n            [secondStep] = \"options.secondStep\"\n            [format]=\"options.format\"\n            [placeholder]=\"options.placeholder\"\n            [use12Hours]=\"options.use12Hours\"\n            (valueChange)=\"onValueChange($event)\"\n            [showHeader]=\"options.showHeader\">\n        </farris-time-picker>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    DatagridTimerComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: Injector }
    ]; };
    DatagridTimerComponent.propDecorators = {
        instance: [{ type: ViewChild, args: ['timepicker',] }]
    };
    return DatagridTimerComponent;
}(DatagridBaseEditorDirective));
export { DatagridTimerComponent };
if (false) {
    /** @type {?} */
    DatagridTimerComponent.prototype.instance;
}
