// @farris/ui-datagrid-editors/esm5/lib/editors/datagrid-sorteditor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Renderer2, ElementRef, NgZone, Injector, ViewChild } from '@angular/core';
import { SorterComponent } from '@farris/ui-sort-editor';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { SortEditorDefaultOptions } from '../editor-default-options';
var DatagridSortEditorComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridSortEditorComponent, _super);
    function DatagridSortEditorComponent(render, el, ngzone, injector) {
        var _this = _super.call(this, render, el, injector) || this;
        _this.ngzone = ngzone;
        _this.injector = injector;
        return _this;
    }
    /**
     * @return {?}
     */
    DatagridSortEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.options = Object.assign({}, SortEditorDefaultOptions, this.options);
    };
    /**
     * @return {?}
     */
    DatagridSortEditorComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.inputElement = this.instance.textbox.nativeElement;
        _super.prototype.ngAfterViewInit.call(this);
    };
    DatagridSortEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'datagrid-sort-editor',
                    template: "\n    <div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup farris-group-auto\">\n        <sorter #sorter\n            style=\"width: 100%\"\n            [formControlName]=\"column.field\"\n            [columns]=\"options.columns\"\n            [fieldEditor]=\"options.fieldEditor\"\n            [enableClear]=\"options.enableClear\"\n            [disabled]=\"options.disabled\"\n            [readonly]=\"options.readonly\"\n            [editable]=\"options.editable\"\n            [returnType]=\"options.returnType\"\n        ></sorter>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    DatagridSortEditorComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: NgZone },
        { type: Injector }
    ]; };
    DatagridSortEditorComponent.propDecorators = {
        instance: [{ type: ViewChild, args: ['sorter',] }]
    };
    return DatagridSortEditorComponent;
}(DatagridBaseEditorDirective));
export { DatagridSortEditorComponent };
if (false) {
    /** @type {?} */
    DatagridSortEditorComponent.prototype.instance;
    /**
     * @type {?}
     * @private
     */
    DatagridSortEditorComponent.prototype.ngzone;
    /** @type {?} */
    DatagridSortEditorComponent.prototype.injector;
}
