// @farris/ui-datagrid-editors/esm5/lib/editors/datagrid-textbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-12 11:10:33
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-10-14 13:19:18
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Renderer2, ElementRef, ViewChild, Injector } from '@angular/core';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { InputDefaultOptions } from '../editor-default-options';
var DatagridTextboxComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DatagridTextboxComponent, _super);
    function DatagridTextboxComponent(render, el, injector) {
        var _this = _super.call(this, render, el, injector) || this;
        _this.inputType = 'text';
        return _this;
    }
    /**
     * @return {?}
     */
    DatagridTextboxComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.options = Object.assign({}, InputDefaultOptions, this.options);
        this.inputElement = this.instance.nativeElement;
        if (this.options.isPassword) {
            this.inputType = 'password';
        }
        else {
            this.inputType = 'text';
        }
    };
    /**
     * @return {?}
     */
    DatagridTextboxComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnDestroy.call(this);
        this.instance = null;
        this.inputElement = null;
    };
    DatagridTextboxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'grid-editor-textbox',
                    template: "\n    <div [formGroup]=\"group\" class=\"f-datagrid-cell-formgroup farris-group-auto\">\n        <input #input [attr.type]=\"inputType\" class=\"form-control \"\n            [attr.placeholder]=\"options?.placeholder\" [formControlName]=\"column.field\"  maxlength=\"{{options?.maxLength}}\">\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    DatagridTextboxComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: Injector }
    ]; };
    DatagridTextboxComponent.propDecorators = {
        instance: [{ type: ViewChild, args: ['input',] }]
    };
    return DatagridTextboxComponent;
}(DatagridBaseEditorDirective));
export { DatagridTextboxComponent };
if (false) {
    /** @type {?} */
    DatagridTextboxComponent.prototype.instance;
    /** @type {?} */
    DatagridTextboxComponent.prototype.inputType;
}
