// @gsp-sys/rtf-apphelp/esm5/lib/rtf-app-help/service/rtf-app.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { RtfUIWebapiService } from '../../common/rtf-ui-webapi.service';
var RtfAppService = /** @class */ (function () {
    function RtfAppService(rtfCommon, httpClient) {
        this.rtfCommon = rtfCommon;
        this.httpClient = httpClient;
    }
    /**
     * 获取功能菜单操作
     */
    /**
     * 获取功能菜单操作
     * @return {?}
     */
    RtfAppService.prototype.getAppList = /**
     * 获取功能菜单操作
     * @return {?}
     */
    function () {
        /** @type {?} */
        var url = this.rtfCommon.functionsApi;
        /** @type {?} */
        var header = this.rtfCommon.httpHeader;
        return this.httpClient.get(url, header);
    };
    /**
     * 获取包含BO数据的应用数据
     */
    /**
     * 获取包含BO数据的应用数据
     * @return {?}
     */
    RtfAppService.prototype.getAppListWithBizObjectData = /**
     * 获取包含BO数据的应用数据
     * @return {?}
     */
    function () {
        /** @type {?} */
        var url = this.rtfCommon.appsWithBOApi;
        /** @type {?} */
        var header = this.rtfCommon.httpHeader;
        return this.httpClient.get(url, header);
    };
    /**
     * 获取多语资源文件
     */
    /**
     * 获取多语资源文件
     * @param {?} url
     * @return {?}
     */
    RtfAppService.prototype.getI18nJson = /**
     * 获取多语资源文件
     * @param {?} url
     * @return {?}
     */
    function (url) {
        /** @type {?} */
        var header = this.rtfCommon.httpHeader;
        return this.httpClient.get(url, header);
    };
    RtfAppService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    RtfAppService.ctorParameters = function () { return [
        { type: RtfUIWebapiService },
        { type: HttpClient }
    ]; };
    return RtfAppService;
}());
export { RtfAppService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    RtfAppService.prototype.rtfCommon;
    /**
     * @type {?}
     * @private
     */
    RtfAppService.prototype.httpClient;
}
