// @farris/ui-batch-edit-dialog/esm5/lib/result-confirm/result-confirm.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
var ResultConfirmComponent = /** @class */ (function () {
    function ResultConfirmComponent() {
        this.dialogRef = null;
        this.rows = 0;
        this.confirmed = new EventEmitter();
        this.canceled = new EventEmitter();
    }
    Object.defineProperty(ResultConfirmComponent.prototype, "showNotify", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var _showNotify = localStorage.getItem('BATCH_EDIT_DIALOG_SHOW_NOTIFY');
            if (_showNotify === 'always' || !_showNotify) {
                return true;
            }
            else {
                return false;
            }
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            /** @type {?} */
            var _showNotify = 'never';
            if (value === true) {
                _showNotify = 'always';
            }
            else {
                _showNotify = 'never';
            }
            localStorage.setItem('BATCH_EDIT_DIALOG_SHOW_NOTIFY', _showNotify);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ResultConfirmComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ResultConfirmComponent.prototype.cancel = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.canceled.emit({ instance: this, event: event });
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ResultConfirmComponent.prototype.confirm = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.confirmed.emit({ instance: this, event: event });
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ResultConfirmComponent.prototype.setRemindOpporunity = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var checked = event && event.target && event.target.checked || false;
        if (checked === true) {
            this.showNotify = false;
        }
        else {
            this.showNotify = true;
        }
    };
    ResultConfirmComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-result-confirm',
                    template: "<div class=\"result-confirm modal-message\">\n  <div class=\"modal-content\">\n    <div class=\"modal-header\">\n      <div class=\"modal-title\">\n        <span class=\"modal-title-label\">{{'batchEditDialog.confirmTitle' | locale:'\u63D0\u793A'}}</span>\n      </div>\n      <div class=\"actions\">\n        <ul>\n          <li class=\"f-btn-icon f-bare\"><span class=\"f-icon modal_close\" (click)='cancel($event)'></span></li>\n        </ul>\n      </div>\n    </div>\n    <div class=\"f-utils-absolute-all\" style=\"height: 10px; z-index: 1; display: none;\"></div>\n    <div class=\"modal-body\">\n      <div style=\"position: static;\" class=\"ps\">\n        <div class=\"ps-content\">\n          <div class=\"farris-messager\">\n            <section class=\"modal-tips d-flex flex-row messager-type-question\">\n              <div class=\"float-left modal-tips-iconwrap\"><span class=\"f-icon f-icon-warning\"></span></div>\n              <div class=\"modal-tips-content\">\n                <p class=\"toast-msg-title\">{{'batchEditDialog.confirmText' | locale:'\u5C06\u4FEE\u6539{0}\u884C\u6570\u636E\uFF0C\u786E\u5B9A\u4FEE\u6539\u5417\uFF1F' | replaceX:\n                  rows}}</p>\n              </div>\n            </section>\n            <div class=\"modal-footer\">\n              <div class=\"custom-control custom-checkbox custom-control-inline never-show\">\n                <input class=\"custom-control-input\" id=\"never\" type=\"checkbox\" (change)=\"setRemindOpporunity($event)\">\n                <label class=\"custom-control-label\" for=\"never\">{{'batchEditDialog.neverShow' | locale:'\u4E0D\u518D\u63D0\u793A'}}</label>\n              </div>\n              <button class=\"btn btn-secondary  btn-lg\" type=\"button\"\n                (click)='cancel($event)'>{{'batchEditDialog.cancelText' | locale:'\u53D6\u6D88'}}</button>\n              <button class=\"btn btn-primary btn-lg\" type=\"button\" (click)=\"confirm($event)\">{{'batchEditDialog.okText'\n                | locale:'\u786E\u5B9A'}}</button>\n            </div>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>",
                    styles: [".result-confirm{height:100%}.result-confirm .modal-header{border-bottom:none}.modal-footer .never-show{text-align:left;width:90%;color:#2d2f33;font-size:14px}"]
                }] }
    ];
    /** @nocollapse */
    ResultConfirmComponent.ctorParameters = function () { return []; };
    ResultConfirmComponent.propDecorators = {
        dialogRef: [{ type: Input }],
        rows: [{ type: Input }],
        confirmed: [{ type: Output }],
        canceled: [{ type: Output }],
        showNotify: [{ type: Input }]
    };
    return ResultConfirmComponent;
}());
export { ResultConfirmComponent };
if (false) {
    /** @type {?} */
    ResultConfirmComponent.prototype.dialogRef;
    /** @type {?} */
    ResultConfirmComponent.prototype.rows;
    /** @type {?} */
    ResultConfirmComponent.prototype.confirmed;
    /** @type {?} */
    ResultConfirmComponent.prototype.canceled;
}
