// @farris/command-services/esm2015/lib/entity-services/entity-traversing.service.js
import { Injectable } from '@angular/core';
import { FrameContext, BindingList } from '@farris/devkit';
/**
 * 实体遍历服务
 * @todo
 * 实体层没有封装相关方法，临时通过BindingData来获取数据
 */
class EntityTraversingService {
    constructor(frameContext) {
        this.frameContext = frameContext;
    }
    get bindingData() {
        return this.frameContext.bindingData;
    }
    /**
     * 获取值
     */
    getPropValue(path) {
        return this.bindingData.getValue(path);
    }
    /**
     * 获取实体数据
     */
    getEntityData(path) {
        let bindingObj;
        const bindingListOrObject = this.bindingData.getValue(path);
        if (bindingListOrObject instanceof BindingList === true) {
            bindingObj = bindingListOrObject.currentItem;
        }
        else {
            bindingObj = bindingListOrObject;
        }
        return bindingObj.toJSON();
    }
    /**
     * 获取实体列表数据
     */
    getEntityListData(path) {
        const bindingList = this.bindingData.getValue(path);
        return bindingList.toJSON();
    }
}
EntityTraversingService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
EntityTraversingService.ctorParameters = () => [
    { type: FrameContext }
];
export { EntityTraversingService };
