// @farris/ui-datepicker/esm5/lib/dynamic-date-input/dynamic-date-input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, HostListener, Injector, Input, Output, ViewChild } from '@angular/core';
var DynamicDateInputComponent = /** @class */ (function () {
    function DynamicDateInputComponent(inject) {
        this.inject = inject;
        this.inputClick = new EventEmitter();
        this.tabKeydown = new EventEmitter();
        this.clear = new EventEmitter();
    }
    /**
     * @return {?}
     */
    DynamicDateInputComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} event
     * @return {?}
     */
    DynamicDateInputComponent.prototype.onMouseEnter = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.value && !this.disabled && !this.readonly) {
            if (this.value && this.clearIcon) {
                this.clearIcon.nativeElement.style.visibility = '';
            }
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DynamicDateInputComponent.prototype.onMouseLeave = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this.disabled && this.clearIcon && !this.readonly) {
            this.clearIcon.nativeElement.style.visibility = 'hidden';
        }
    };
    DynamicDateInputComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dynamic-date-input',
                    template: "<div class=\"input-group\" [ngClass]=\"{\r\n    'f-state-disabled': disabled,\r\n    'f-state-readonly': readonly && !disabled,\r\n    'f-state-focus':isFocus\r\n}\">\r\n    <input type=\"text\" class=\"form-control f-utils-fill\"  (click)=\"inputClick.emit($event)\"\r\n    [attr.placeholder]=\"readonly||disabled?'':placeholder\" \r\n    name=\"farris-dynamic-date\" [ngModel]=\"text\"\r\n    #dynamicDateInput autocomplete=\"off\" [readonly]=\"true\"\r\n    [disabled]=\"disabled\" \r\n    style=\"padding-right: 4px;\" (keydown.tab)=\"tabKeydown.emit($event)\"  />\r\n\r\n    <div class=\"input-group-append f-cmp-iconbtn-wrapper\" *ngIf=\"!disabled\" style=\"position: relative;\">\r\n        <span class=\"input-group-text input-group-clear datepicker-clear\" (click)=\"clear.emit($event)\"\r\n            #clearIcon *ngIf=\"showClear\" style=\"position: absolute;height: 100%;right: 0px;top: 0px;z-index: 100;visibility: hidden\">\r\n            <i class=\"f-icon modal_close\"></i>\r\n        </span>\r\n        <button type=\"button\" class=\"btn f-cmp-iconbtn\" (click)=\"inputClick.emit($event)\">\r\n            <span class=\"f-icon f-icon-date\"></span>\r\n        </button>\r\n    </div>\r\n</div>"
                }] }
    ];
    /** @nocollapse */
    DynamicDateInputComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    DynamicDateInputComponent.propDecorators = {
        readonly: [{ type: Input }],
        disabled: [{ type: Input }],
        value: [{ type: Input }],
        text: [{ type: Input }],
        isFocus: [{ type: Input }],
        placeholder: [{ type: Input }],
        showClear: [{ type: Input }],
        inputClick: [{ type: Output }],
        tabKeydown: [{ type: Output }],
        clear: [{ type: Output }],
        clearIcon: [{ type: ViewChild, args: ['clearIcon',] }],
        onMouseEnter: [{ type: HostListener, args: ['mouseenter', ['$event'],] }],
        onMouseLeave: [{ type: HostListener, args: ['mouseleave', ['$event'],] }]
    };
    return DynamicDateInputComponent;
}());
export { DynamicDateInputComponent };
if (false) {
    /** @type {?} */
    DynamicDateInputComponent.prototype.readonly;
    /** @type {?} */
    DynamicDateInputComponent.prototype.disabled;
    /** @type {?} */
    DynamicDateInputComponent.prototype.value;
    /** @type {?} */
    DynamicDateInputComponent.prototype.text;
    /** @type {?} */
    DynamicDateInputComponent.prototype.isFocus;
    /** @type {?} */
    DynamicDateInputComponent.prototype.placeholder;
    /** @type {?} */
    DynamicDateInputComponent.prototype.showClear;
    /** @type {?} */
    DynamicDateInputComponent.prototype.inputClick;
    /** @type {?} */
    DynamicDateInputComponent.prototype.tabKeydown;
    /** @type {?} */
    DynamicDateInputComponent.prototype.clear;
    /** @type {?} */
    DynamicDateInputComponent.prototype.clearIcon;
    /**
     * @type {?}
     * @private
     */
    DynamicDateInputComponent.prototype.inject;
}
