// @farris/ui-tooltip/esm5/lib/attention.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, RendererFactory2, NgZone } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function AttentionElement() { }
if (false) {
    /** @type {?} */
    AttentionElement.prototype.idOrEl;
    /** @type {?|undefined} */
    AttentionElement.prototype.type;
    /** @type {?|undefined} */
    AttentionElement.prototype.class;
}
var AttentionService = /** @class */ (function () {
    function AttentionService(renderFac, ngZone) {
        this.ngZone = ngZone;
        this.renderer = renderFac.createRenderer(null, null);
    }
    /**
     * @param {?} els
     * @param {?=} timer
     * @return {?}
     */
    AttentionService.prototype.catchAttention = /**
     * @param {?} els
     * @param {?=} timer
     * @return {?}
     */
    function (els, timer) {
        var _this = this;
        if (timer === void 0) { timer = 600; }
        if (els && els.length > 0) {
            /** @type {?} */
            var attentionEls_1 = [];
            for (var k = 0; k < els.length; k++) {
                /** @type {?} */
                var attentionEl;
                if (typeof els[k].idOrEl == 'string') {
                    attentionEl = document.getElementById(els[k].idOrEl);
                }
                else {
                    attentionEl = els[k].idOrEl;
                }
                if (attentionEl) {
                    attentionEls_1.push(attentionEl);
                    this.renderer.addClass(attentionEl, 'f-catch-attention-debounce');
                }
            }
            // 移除class
            this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            function () {
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    for (var m = 0; m < attentionEls_1.length; m++) {
                        _this.renderer.removeClass(attentionEls_1[m], 'f-catch-attention-debounce');
                    }
                }), timer);
            }));
        }
    };
    /**
     * @return {?}
     */
    AttentionService.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    AttentionService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AttentionService.ctorParameters = function () { return [
        { type: RendererFactory2 },
        { type: NgZone }
    ]; };
    /** @nocollapse */ AttentionService.ngInjectableDef = i0.defineInjectable({ factory: function AttentionService_Factory() { return new AttentionService(i0.inject(i0.RendererFactory2), i0.inject(i0.NgZone)); }, token: AttentionService, providedIn: "root" });
    return AttentionService;
}());
export { AttentionService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AttentionService.prototype.renderer;
    /** @type {?} */
    AttentionService.prototype.ngZone;
}
