// @farris/ui-tooltip/esm5/lib/tooltip.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { FarrisTooltipDirective } from "./tooltip.directive";
import { FInputTipsDirective } from "./inputTips.directive";
import { TooltipComponent } from "./tooltip-component/tooltip.component";
import { AttentionService } from "./attention.service";
import { tooltipHtmlPipe } from "./tooltip-safe-html.pipe";
import { FarrisCommonModule } from "@farris/ui-common";
var FarrisTooltipModule = /** @class */ (function () {
    function FarrisTooltipModule() {
    }
    FarrisTooltipModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        FarrisTooltipDirective,
                        TooltipComponent,
                        FInputTipsDirective,
                        tooltipHtmlPipe,
                    ],
                    imports: [
                        CommonModule,
                        FarrisCommonModule
                    ],
                    providers: [AttentionService],
                    entryComponents: [TooltipComponent],
                    exports: [FarrisTooltipDirective, TooltipComponent, FInputTipsDirective],
                },] }
    ];
    return FarrisTooltipModule;
}());
export { FarrisTooltipModule };
