// @farris/ui-tooltip/esm5/lib/tooltip.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function PlacementInfo() { }
if (false) {
    /** @type {?} */
    PlacementInfo.prototype.placement;
    /** @type {?} */
    PlacementInfo.prototype.top;
    /** @type {?} */
    PlacementInfo.prototype.left;
}
var TooltipService = /** @class */ (function () {
    function TooltipService() {
        /* tooltip位置 */
        this.placement$ = new Subject();
        /* tooltip控制显示 */
        this.show$ = new Subject();
        /* 计算位置 */
        this.calPosition$ = new Subject();
    }
    /**
     * @return {?}
     */
    TooltipService.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    TooltipService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TooltipService.ctorParameters = function () { return []; };
    /** @nocollapse */ TooltipService.ngInjectableDef = i0.defineInjectable({ factory: function TooltipService_Factory() { return new TooltipService(); }, token: TooltipService, providedIn: "root" });
    return TooltipService;
}());
export { TooltipService };
if (false) {
    /** @type {?} */
    TooltipService.prototype.placement$;
    /** @type {?} */
    TooltipService.prototype.show$;
    /** @type {?} */
    TooltipService.prototype.calPosition$;
}
