// @gsp-wf/rtdevkit/esm5/lib/submit/process-selector/process-selector.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, Optional } from '@angular/core';
import { Subject } from 'rxjs';
import { MessagerService } from '@farris/ui-messager';
import { I18nService } from '../../runtime/services/i18n.service';
var ProcessSelectorComponent = /** @class */ (function () {
    function ProcessSelectorComponent(msgService, i18nService) {
        this.msgService = msgService;
        this.i18nService = i18nService;
        this.cls = 'farris-main-area flex-column';
        this.columns = [
            { title: this.i18nService ? this.i18nService.getValue('static.columnName.prcessName') : '流程名称', field: 'procDefName', width: 200 },
            { title: this.i18nService ? this.i18nService.getValue('static.columnName.version') : '版本', field: 'version', width: 100 }
        ];
        this.idField = 'procDefId';
        this.mySelection = [];
        this.subject = new Subject();
    }
    /**
     * @return {?}
     */
    ProcessSelectorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    ProcessSelectorComponent.prototype.btnOKClick = /**
     * @return {?}
     */
    function () {
        if (this.mySelection.length < 1) {
            this.msgService.warning(this.i18nService ? this.i18nService.getValue('static.submit.noProcess') : '尚未选择流程定义，请了解。');
            return;
        }
        this.subject.next(this.mySelection[0]);
    };
    /**
     * @return {?}
     */
    ProcessSelectorComponent.prototype.btnCancelClick = /**
     * @return {?}
     */
    function () {
        this.subject.next(null);
    };
    /**
     * @return {?}
     */
    ProcessSelectorComponent.prototype.subscribe = /**
     * @return {?}
     */
    function () {
        return this.subject.asObservable();
    };
    /**
     * @param {?} data
     * @return {?}
     */
    ProcessSelectorComponent.prototype.onSelectionChange = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data && data.data) {
            this.mySelection = [data.data.procDefId];
        }
        else {
            this.mySelection = [];
        }
    };
    ProcessSelectorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-process-selector',
                    template: "<div style=\"background-color: white; overflow: auto;\">\r\n    <farris-datatable class=\"f-utils-fill\" [data]=\"gridData\" [idField]=\"idField\" [columns]=\"columns\" [pagination]=\"false\" [showFilterBar]='false'\r\n        [singleSelect]=\"true\" (selectedRow)=\"onSelectionChange($event)\" [fill]=\"true\" [height]=\"250\">\r\n    </farris-datatable>\r\n</div>\r\n<div class=\"modal-footer flex-shrink-0\">\r\n    <button type=\"button\" class=\"btn btn-primary\" (click)=\"btnOKClick()\">{{('static.footOk' | translate)?('static.footOk' | translate):'\u786E\u5B9A' }}</button>\r\n    <button type=\"button\" class=\"btn btn-default\" (click)=\"btnCancelClick()\">{{('static.footCancel' | translate)?('static.footCancel' | translate):'\u53D6\u6D88' }}</button>\r\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ProcessSelectorComponent.ctorParameters = function () { return [
        { type: MessagerService },
        { type: I18nService, decorators: [{ type: Optional }] }
    ]; };
    ProcessSelectorComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }]
    };
    return ProcessSelectorComponent;
}());
export { ProcessSelectorComponent };
if (false) {
    /** @type {?} */
    ProcessSelectorComponent.prototype.cls;
    /** @type {?} */
    ProcessSelectorComponent.prototype.gridData;
    /** @type {?} */
    ProcessSelectorComponent.prototype.columns;
    /** @type {?} */
    ProcessSelectorComponent.prototype.idField;
    /** @type {?} */
    ProcessSelectorComponent.prototype.mySelection;
    /**
     * @type {?}
     * @private
     */
    ProcessSelectorComponent.prototype.subject;
    /**
     * @type {?}
     * @private
     */
    ProcessSelectorComponent.prototype.msgService;
    /**
     * @type {?}
     * @private
     */
    ProcessSelectorComponent.prototype.i18nService;
}
