// @edp-aif/runtime-api/esm5/lib/aif-manual-creation/components/source-data-selector/source-data-selector.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Inject, LOCALE_ID } from '@angular/core';
import { Subject } from 'rxjs';
import { TreeTableComponent } from '@farris/ui-treetable';
/**
 * @deprecated please use \@edp-bif/runtime-api
 */
var SourceDataSelectorComponent = /** @class */ (function () {
    function SourceDataSelectorComponent(localeId) {
        var _this = this;
        this.subject4EntityDatas = new Subject();
        this.cols = [];
        this.treedata = [];
        this.clickOk4AifSelector = (/**
         * @return {?}
         */
        function () {
            _this.subject4EntityDatas.next([]);
            _this.subject4EntityDatas.complete();
            _this.bsModalRef4AifSelector.close();
        });
        this.clickClose4AifSelector = (/**
         * @return {?}
         */
        function () {
            _this.bsModalRef4AifSelector.close();
        });
    }
    /**
     * @return {?}
     */
    SourceDataSelectorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    SourceDataSelectorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-source-data-selector',
                    template: "<div class=\"g-main p-2\">\r\n  <farris-treetable #treetable [columns]=\"cols\" [showBorder]=\"false\" [showHeader]=\"false\" [data]=\"treedata\"\r\n    [idField]=\"'id'\" [fit]=\"true\" [singleSelect]=\"true\" [showCheckbox]=\"false\" [showIcon]=\"true\" [disabled]=\"false\">\r\n  </farris-treetable>\r\n\r\n  <ng-template #cell let-ctx>\r\n\r\n  </ng-template>\r\n\r\n  <p>\r\n    Hello world\r\n  </p>\r\n\r\n  <!-- <div class=\"form-group\" style=\"float: right;\">\r\n    <button class=\"btn-primary btn btn-sm ml-2\" (click)=\"clickOk4Aif()\">\u786E\u5B9A</button>\r\n    <button class=\"btn btn-sm ml-2 btn-default\" (click)=\"clickClose4Aif()\">\u53D6\u6D88</button>\r\n  </div> -->\r\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    SourceDataSelectorComponent.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    SourceDataSelectorComponent.propDecorators = {
        treetable: [{ type: ViewChild, args: ["treetable",] }]
    };
    return SourceDataSelectorComponent;
}());
export { SourceDataSelectorComponent };
if (false) {
    /** @type {?} */
    SourceDataSelectorComponent.prototype.bsModalRef4AifSelector;
    /** @type {?} */
    SourceDataSelectorComponent.prototype.subject4EntityDatas;
    /** @type {?} */
    SourceDataSelectorComponent.prototype.treetable;
    /** @type {?} */
    SourceDataSelectorComponent.prototype.cols;
    /** @type {?} */
    SourceDataSelectorComponent.prototype.treedata;
    /**
     * @type {?}
     * @private
     */
    SourceDataSelectorComponent.prototype.langCode;
    /** @type {?} */
    SourceDataSelectorComponent.prototype.clickOk4AifSelector;
    /** @type {?} */
    SourceDataSelectorComponent.prototype.clickClose4AifSelector;
}
