// @farris/ui-response-toolbar/esm2015/lib/resized/f-resized.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Output, EventEmitter, ElementRef, HostListener } from '@angular/core';
import { Subject } from 'rxjs';
import { throttleTime } from 'rxjs/operators';
import { ResizedEvent } from './f-resized.events';
export class FResizedDirective {
    /**
     * @param {?} element
     */
    constructor(element) {
        this.element = element;
        this.resizedEvent = new EventEmitter();
        this.resizeSubject = new Subject();
        this.resizeObservable = this.resizeSubject
            .asObservable()
            .pipe(throttleTime(200));
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onResize(event) {
        this.resizeSubject.next(event);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.resizeObservable.subscribe((/**
         * @param {?} x
         * @return {?}
         */
        x => this.doWithResize()));
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.doWithResize();
    }
    /**
     * @private
     * @return {?}
     */
    doWithResize() {
        /** @type {?} */
        const newWidth = this.element.nativeElement.clientWidth;
        /** @type {?} */
        const newHeight = this.element.nativeElement.clientHeight;
        if (newWidth === this.oldWidth && newHeight === this.oldHeight) {
            return;
        }
        this.resizedEvent.next(new ResizedEvent(this.element, newWidth, newHeight, this.oldWidth, this.oldHeight));
        this.oldWidth = newWidth;
        this.oldHeight = newHeight;
    }
}
FResizedDirective.decorators = [
    { type: Directive, args: [{
                selector: '[fResized]'
            },] }
];
/** @nocollapse */
FResizedDirective.ctorParameters = () => [
    { type: ElementRef }
];
FResizedDirective.propDecorators = {
    resizedEvent: [{ type: Output }],
    onResize: [{ type: HostListener, args: ['window:resize', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    FResizedDirective.prototype.resizedEvent;
    /**
     * @type {?}
     * @private
     */
    FResizedDirective.prototype.oldWidth;
    /**
     * @type {?}
     * @private
     */
    FResizedDirective.prototype.oldHeight;
    /**
     * @type {?}
     * @private
     */
    FResizedDirective.prototype.resizeSubject;
    /**
     * @type {?}
     * @private
     */
    FResizedDirective.prototype.resizeObservable;
    /**
     * @type {?}
     * @private
     */
    FResizedDirective.prototype.element;
}
