// @gsp-cmp/ubcc-param-data-ui/esm5/lib/controls/param-control/input-control/input-control.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/controls/param-control/input-control/input-control.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var InputControlComponent = /** @class */ (function () {
    function InputControlComponent() {
        this.canEdit = true;
    }
    /**
     * @return {?}
     */
    InputControlComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.param = this.param || {};
    };
    /**
     * @return {?}
     */
    InputControlComponent.prototype.isReadonly = /**
     * @return {?}
     */
    function () {
        if (!this.canEdit) {
            return true;
        }
        else {
            return !this.param || !this.param.readonly ? false : true;
        }
    };
    /**
     * @return {?}
     */
    InputControlComponent.prototype.show = /**
     * @return {?}
     */
    function () {
        return !this.param || !this.param.description ? false : true;
    };
    InputControlComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-input-control',
                    template: "<div class=\"farris-form-controls-inline farris-group-wrap col-12\">\n  <div class=\"form-group farris-form-group\">\n    <label class=\"col-form-label col-4\">\n      <span class=\"farris-label-text\" style=\"float:right\">{{param.name}}</span>\n    </label>\n    <div class=\"farris-input-wrap col-6\">\n      <input-group type=\"text\" placeholder=\"\u8BF7\u8F93\u5165\" [readonly]=\"isReadonly()\" autocomplete=\"off\" [(value)]=\"param.value\"></input-group>\n    </div>\n    <span *ngIf=\"show()\" farrisTooltip [placement]=\"'right-top'\" [content]=\"param.description\" class=\"f-icon f-icon-info\"></span>\n  </div>\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    InputControlComponent.ctorParameters = function () { return []; };
    InputControlComponent.propDecorators = {
        param: [{ type: Input }],
        canEdit: [{ type: Input }]
    };
    return InputControlComponent;
}());
export { InputControlComponent };
if (false) {
    /** @type {?} */
    InputControlComponent.prototype.param;
    /** @type {?} */
    InputControlComponent.prototype.canEdit;
}
