// @gsp-wf/rtdevkit/esm5/lib/runtime/components/task-signadd/task-signadd.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { UserHelperComponent } from '../user-helper/user-helper.component';
import { TaskService } from '../../services/task.service';
import { UserService } from '../../services/user.service';
import { NotifyService } from '@farris/ui-notify';
import { I18nService } from '../../services/i18n.service';
var TaskSignaddComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaskSignaddComponent, _super);
    function TaskSignaddComponent(taskSvc, userSvc, notify, i18nService) {
        var _this = _super.call(this, i18nService, userSvc) || this;
        _this.taskSvc = taskSvc;
        _this.userSvc = userSvc;
        _this.notify = notify;
        _this.i18nService = i18nService;
        _this.completed = new EventEmitter();
        _this.paginationOptions.id = 'WF-SignAdd-DataTable-Pagination';
        return _this;
    }
    Object.defineProperty(TaskSignaddComponent.prototype, "addsign_context", {
        set: /**
         * @param {?} context
         * @return {?}
         */
        function (context) {
            this.selectedUsers = [];
            this.addSignContext = tslib_1.__assign({}, context, { assignees: [] });
            this.showModal(!!this.addSignContext.workItemId);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TaskSignaddComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    TaskSignaddComponent.prototype.handleOk = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.selectedUsers = this.userTable.selections;
        // 征询
        if (this.addSignContext.action === 'AddSignFront') {
            // 需要校验选中用户
            if (!this.selectedUsers || this.selectedUsers.length < 1) {
                this.notify.info(this.i18nService.getValue('static.addsign.selectConsultant'));
                return;
            }
            this.addSignContext.assignees = this.selectedUsers.map((/**
             * @param {?} user
             * @return {?}
             */
            function (user) { return user.ID; }));
            this.taskSvc.addSign(this.addSignContext).subscribe((/**
             * @return {?}
             */
            function () {
                _this.notify.success(_this.i18nService.getValue('static.addsign.successConsult'));
                _this.completed.emit(true);
                _this.showModal(false);
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) {
                if (error && error.error && error.error.Message) {
                    _this.notify.error(error.error.Message);
                }
            }));
        }
        else {
            // 加签
            this.completed.emit(true);
            this.showModal(false);
        }
    };
    TaskSignaddComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-task-signadd',
                    template: "<farris-dialog #allUserDialog [buttons]=\"buttonRef\" [width]=\"850\" [height]=\"520\" [title]=\"'\u7CFB\u7EDF\u7528\u6237'\">\r\n  <div class=\"d-flex flex-column my-data-grid h-100\" style=\"padding: 0 14px;\">\r\n    <div class=\"flex-auto\" style=\"position:relative;\">\r\n      <farris-datatable #allUserTable [columns]=\"columns\" [data]=\"userListData | paginate: paginationOptions\"\r\n        [showFilterBar]='true' [singleSelect]=\"singleSelect\" [idField]=\"'ID'\" [remote]=\"'server'\" (search)=\"search($event)\"\r\n        [pagination]='false'></farris-datatable>\r\n    </div>\r\n    <div style=\"position:relative;\">\r\n      <pagination-controls [id]=\"paginationOptions.id\" [previousLabel]=\"' '\" [nextLabel]=\"' '\"\r\n        (pageChange)=\"onPageChange($event)\" (pageSizeChange)=\"onPageSizeChange($event)\">\r\n      </pagination-controls>\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n<ng-template #buttonRef>\r\n  <button class=\"btn btn-sm btn-primary px-3\" (click)=\"handleOk()\">\u786E\u5B9A</button>\r\n  <button class=\"btn btn-sm btn-default px-3\" (click)=\"handleCancel()\">\u53D6\u6D88</button>\r\n</ng-template>",
                    styles: [".my-data-grid .filter-row .search-fields{padding-left:0!important}.my-data-grid .filter-row .search-text{padding-right:0!important}"]
                }] }
    ];
    /** @nocollapse */
    TaskSignaddComponent.ctorParameters = function () { return [
        { type: TaskService },
        { type: UserService },
        { type: NotifyService },
        { type: I18nService }
    ]; };
    TaskSignaddComponent.propDecorators = {
        addsign_context: [{ type: Input }],
        completed: [{ type: Output }]
    };
    return TaskSignaddComponent;
}(UserHelperComponent));
export { TaskSignaddComponent };
if (false) {
    /** @type {?} */
    TaskSignaddComponent.prototype.completed;
    /** @type {?} */
    TaskSignaddComponent.prototype.addSignContext;
    /**
     * @type {?}
     * @private
     */
    TaskSignaddComponent.prototype.taskSvc;
    /**
     * @type {?}
     * @protected
     */
    TaskSignaddComponent.prototype.userSvc;
    /**
     * @type {?}
     * @protected
     */
    TaskSignaddComponent.prototype.notify;
    /**
     * @type {?}
     * @protected
     */
    TaskSignaddComponent.prototype.i18nService;
}
