// @farris/ui-filter/esm2015/lib/entity/conditionvalue/flexiblenumbervalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
export class FlexibleNumberValue {
    /**
     * @param {?=} value
     * @param {?=} single
     */
    constructor(value = { startValue: null, endValue: null, value: null }, single = true) {
        this.clear$ = new EventEmitter();
        this.startValue = value.startValue;
        this.endValue = value.endValue;
        this.value = value.value;
        this.single = single;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.FlexibleNumber;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        if (this.single) {
            return !this.value;
        }
        else
            return (this.startValue == null && this.endValue == null);
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.startValue = undefined;
        this.endValue = undefined;
        this.value = undefined;
        this.clear$.emit();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        throw new Error("Method not implemented.");
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return {
            begin: this.startValue,
            end: this.endValue,
            value: this.value
        };
    }
    /**
     * @return {?}
     */
    getValueText() {
        if (this.single) {
            if (this.value == undefined || (!this.value && typeof (this.value) != 'undefined' && this.value != 0)) {
                return this.value;
            }
            else {
                return '' + this.value;
            }
        }
        else {
            if (this.startValue == undefined && this.endValue == undefined) {
                return '';
            }
            else if (this.startValue == undefined) {
                return ('~' + this.endValue);
            }
            else if (this.endValue == undefined) {
                return (this.startValue + '~');
            }
            else {
                return (this.startValue + '~' + this.endValue);
            }
        }
    }
}
if (false) {
    /** @type {?} */
    FlexibleNumberValue.prototype.startValue;
    /** @type {?} */
    FlexibleNumberValue.prototype.endValue;
    /** @type {?} */
    FlexibleNumberValue.prototype.value;
    /** @type {?} */
    FlexibleNumberValue.prototype.single;
    /** @type {?} */
    FlexibleNumberValue.prototype.clear$;
}
