// @farris/ui-filter/esm2015/lib/entity/conditionvalue/helpvalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
import { cloneDeep } from 'lodash-es';
export class HelpValue {
    /**
     * @param {?=} data
     */
    constructor(data = { value: [], valueField: undefined, textValue: undefined }) {
        this.clear$ = new EventEmitter();
        if (data.hasOwnProperty('textValue')) { // 新结构，value是数组
            this.valueField = data.valueField;
            this.value = data.value;
            this.textValue = data.textValue;
        }
        else { // 旧结构，value是对象
            this.valueField = data.valueField;
            this.value = data.value ? [data.value] : [];
            this.textValue = 'help-text-value-null'; // 旧结构无textValue属性
        }
    }
    /**
     * @param {?} helpItem
     * @param {?} args
     * @return {?}
     */
    static getPropValue(helpItem, args) {
        if (args.length > 1) {
            /** @type {?} */
            const arg = args.shift();
            return helpItem[arg] ? this.getPropValue(helpItem[arg], args) : null;
        }
        return helpItem[args[0]];
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.value = [];
        this.valueField = undefined;
        this.textValue = undefined;
        this.clear$.emit();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        throw new Error("Method not implemented.");
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        if (this.valueField) {
            /** @type {?} */
            const args = this.valueField.split('.');
            /** @type {?} */
            const valueArr = this.value.map((/**
             * @param {?} helpItem
             * @return {?}
             */
            helpItem => HelpValue.getPropValue(helpItem, cloneDeep(args))));
            return valueArr && valueArr.length ? valueArr.join(',') : '';
        }
        else {
            return '';
        }
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.Help;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return !this.valueField;
    }
    /**
     * @param {?} textField
     * @return {?}
     */
    getTextValue(textField) {
        /** @type {?} */
        const args = textField.split('.');
        /** @type {?} */
        const textArr = this.value.map((/**
         * @param {?} helpItem
         * @return {?}
         */
        helpItem => HelpValue.getPropValue(helpItem, cloneDeep(args))));
        return textArr && textArr.length ? textArr.join(',') : '';
    }
    /**
     * @return {?}
     */
    getValueText() {
        return this.textValue;
    }
}
if (false) {
    /** @type {?} */
    HelpValue.prototype.valueField;
    /** @type {?} */
    HelpValue.prototype.value;
    /** @type {?} */
    HelpValue.prototype.textValue;
    /** @type {?} */
    HelpValue.prototype.clear$;
}
