// @farris/ui-filter/esm2015/lib/entity/conditionvalue/numbervalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
export class NumberValue {
    /**
     * @param {?=} value
     */
    constructor(value = { value: null }) {
        this.clear$ = new EventEmitter();
        this.value = value.value;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.Number;
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.value = undefined;
        this.clear$.emit();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        throw new Error("Method not implemented.");
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.value;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return !this.value;
    }
    /**
     * @return {?}
     */
    getValueText() {
        if (this.value == undefined || (!this.value && typeof (this.value) != 'undefined' && this.value != 0)) {
            return this.value;
        }
        else {
            return '' + this.value;
        }
    }
}
if (false) {
    /** @type {?} */
    NumberValue.prototype.value;
    /** @type {?} */
    NumberValue.prototype.clear$;
}
