// @gspwidget/gridster/esm5/lib/gridster/gridster.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { GridsterComponent } from './gridster.component';
import { GridsterItemComponent } from './gridsterItem.component';
import { GridsterPreviewComponent } from './gridsterPreview.component';
var GridsterModule = /** @class */ (function () {
    function GridsterModule() {
    }
    GridsterModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        GridsterComponent,
                        GridsterItemComponent,
                        GridsterPreviewComponent
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [GridsterComponent, GridsterItemComponent],
                    providers: [],
                    bootstrap: []
                },] }
    ];
    return GridsterModule;
}());
export { GridsterModule };
