// @farris/ui-section/esm5/lib/directive/section-toolbar.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, TemplateRef, Input } from '@angular/core';
var FarrisSectionToolbarDirective = /** @class */ (function () {
    function FarrisSectionToolbarDirective(templateRef) {
        this.templateRef = templateRef;
    }
    FarrisSectionToolbarDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farrisSectionToolbar]',
                },] }
    ];
    /** @nocollapse */
    FarrisSectionToolbarDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    FarrisSectionToolbarDirective.propDecorators = {
        toolbarCls: [{ type: Input }]
    };
    return FarrisSectionToolbarDirective;
}());
export { FarrisSectionToolbarDirective };
if (false) {
    /** @type {?} */
    FarrisSectionToolbarDirective.prototype.toolbarCls;
    /** @type {?} */
    FarrisSectionToolbarDirective.prototype.templateRef;
}
var FarrisSectionViewChangeDirective = /** @class */ (function () {
    function FarrisSectionViewChangeDirective(templateRef) {
        this.templateRef = templateRef;
    }
    FarrisSectionViewChangeDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farrisSectionViewChange]',
                },] }
    ];
    /** @nocollapse */
    FarrisSectionViewChangeDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    FarrisSectionViewChangeDirective.propDecorators = {
        viewChangeCls: [{ type: Input }]
    };
    return FarrisSectionViewChangeDirective;
}());
export { FarrisSectionViewChangeDirective };
if (false) {
    /** @type {?} */
    FarrisSectionViewChangeDirective.prototype.viewChangeCls;
    /** @type {?} */
    FarrisSectionViewChangeDirective.prototype.templateRef;
}
