// @gsp-wf/ui-comment/esm2015/lib/services/task-comment.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
export class TaskCommentService {
    /**
     * @param {?} httpSvc
     */
    constructor(httpSvc) {
        this.httpSvc = httpSvc;
    }
    /**
     * @return {?}
     */
    static newId() {
        /** @type {?} */
        const S4 = () => (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
        return (S4() + S4() + '-' + S4() + '-' + S4() + '-' + S4() + '-' + S4() + S4() + S4());
    }
    /**
     * @return {?}
     */
    getCommentList() {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/commonComments`;
        return this.httpSvc.get(url);
    }
    /**
     * @return {?}
     */
    addComment() {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/commonComments`;
        return this.httpSvc.post(url, {});
    }
    /**
     * 获取任务中心配置项
     * @param {?} code 配置项的Code
     * @return {?}
     */
    getTaskCenterConfigByCode(code) {
        /** @type {?} */
        const url = `/api/runtime/task/v1.0/task-center-setting/code?code=${code}`;
        return this.httpSvc.get(url);
    }
    /**
     * 真正上传附件
     * @param {?} files
     * @return {?}
     */
    uploadAttachment(files) {
        /** @type {?} */
        let url = `/api/runtime/dfs/v1.0/doc/commit`;
        return this.httpSvc.put(url, files);
    }
    /**
     * 删除附件
     * @param {?} fileDeleteEntity
     * @return {?}
     */
    deleteAttachment(fileDeleteEntity) {
        /** @type {?} */
        let url = `/api/runtime/dfs/v1.0/doc`;
        return this.httpSvc.delete(url, fileDeleteEntity);
    }
}
TaskCommentService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
TaskCommentService.ctorParameters = () => [
    { type: HttpService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaskCommentService.prototype.httpSvc;
}
