// @qdp/ide-cmp/esm5/lib/component/schema-editor/style-editor/style-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Injector } from '@angular/core';
import { LocalizeService } from '@qdp/localize';
var StyleEditorComponent = /** @class */ (function () {
    function StyleEditorComponent(injector) {
        this.injector = injector;
        this.rowHeight = 35;
        this.fontFamilyData = [
            { name: '宋体', value: 0 },
            { name: '新宋体', value: 1 },
            { name: '仿宋', value: 2 },
            { name: '楷体', value: 3 },
            { name: '黑体', value: 4 },
            { name: 'ARIAL', value: 5 },
            { name: 'ROMAN', value: 6 },
            { name: 'TIMES NEW ROMAN', value: 7 },
            { name: 'VERDANA', value: 8 },
            { name: 'Microsoft YaHei', value: 9 }
        ];
        this.fontFamily = 9;
        this.fontSizeData = [
            { name: 8, value: 8 }, { name: 9, value: 9 }, { name: 10, value: 10 }, { name: 11, value: 11 }, { name: 12, value: 12 },
            { name: 13, value: 13 }, { name: 14, value: 14 }, { name: 15, value: 15 }, { name: 16, value: 16 }, { name: 17, value: 17 },
            { name: 18, value: 18 }, { name: 19, value: 19 }, { name: 20, value: 20 }, { name: 21, value: 21 }, { name: 22, value: 22 },
            { name: 23, value: 23 }, { name: 24, value: 24 }, { name: 25, value: 25 }, { name: 26, value: 26 }, { name: 27, value: 27 },
            { name: 28, value: 28 }, { name: 29, value: 29 }, { name: 30, value: 30 }, { name: 31, value: 31 }, { name: 32, value: 32 },
            { name: 33, value: 33 }, { name: 34, value: 34 }, { name: 35, value: 35 }, { name: 36, value: 36 }, { name: 37, value: 37 },
            { name: 38, value: 38 }, { name: 39, value: 39 }, { name: 40, value: 40 }
        ];
        // fontSizeData: any = [
        //   8, 9, 10, 11, 12, 13, 14, 15, 16, 17,
        //   18, 19, 20, 21, 22, 23, 24, 25, 26,
        //   27, 28, 29, 30, 31, 32, 33, 34, 35,
        //   36, 37, 38, 39, 40
        // ];
        this.fontSize = 14;
        this.fontColor = 'black';
        this.isWeight = false;
        this.isStyle = false;
        this.isUnderLine = false;
        this.fontWeight = 'normal';
        this.fontStyle = 'normal';
        this.fontUnderLine = 'none';
        this.titleActive = true;
        this.subTitleActive = false;
        this.headerActive = false;
        this.dataAreaActive = false;
        this.footerActive = false;
        this.localizeService = this.injector.get(LocalizeService);
        this.fontFamilyData = this.localizeService.getValue('ideCmp.schemaManager.styleOption.fontFamilyData');
    }
    /**
     * @return {?}
     */
    StyleEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    StyleEditorComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        for (var changeName in changes) {
            /** @type {?} */
            var newValue = changes[changeName].currentValue;
            if (newValue !== null && newValue !== void 0) {
                switch (changeName) {
                    case 'currSchema':
                        if (newValue.schemaValue && newValue.schemaValue.styleOption) {
                            /** @type {?} */
                            var str = JSON.stringify(newValue.schemaValue.styleOption);
                            this.styleOption = JSON.parse(str);
                            if (this.styleOption) {
                                if (this.titleActive) {
                                    this.initStyleOption('Title');
                                }
                                if (this.subTitleActive) {
                                    this.initStyleOption('SubTitle');
                                }
                                if (this.headerActive) {
                                    this.initStyleOption('Header');
                                }
                                if (this.dataAreaActive) {
                                    this.initStyleOption('DataArea');
                                }
                                if (this.footerActive) {
                                    this.initStyleOption('Footer');
                                }
                            }
                        }
                        break;
                }
            }
        }
    };
    /**
     * @return {?}
     */
    StyleEditorComponent.prototype.handleTitleBtn = /**
     * @return {?}
     */
    function () {
        this.updateStyleOption();
        this.initStyleOption('Title');
        this.titleActive = true;
        this.subTitleActive = this.headerActive = this.dataAreaActive = this.footerActive = false;
    };
    /**
     * @return {?}
     */
    StyleEditorComponent.prototype.handleSubTitleBtn = /**
     * @return {?}
     */
    function () {
        this.updateStyleOption();
        this.initStyleOption('SubTitle');
        this.subTitleActive = true;
        this.titleActive = this.headerActive = this.dataAreaActive = this.footerActive = false;
    };
    /**
     * @return {?}
     */
    StyleEditorComponent.prototype.handleHeaderBtn = /**
     * @return {?}
     */
    function () {
        this.updateStyleOption();
        this.initStyleOption('Header');
        this.headerActive = true;
        this.titleActive = this.subTitleActive = this.dataAreaActive = this.footerActive = false;
    };
    /**
     * @return {?}
     */
    StyleEditorComponent.prototype.handleDataAreaBtn = /**
     * @return {?}
     */
    function () {
        this.updateStyleOption();
        this.initStyleOption('DataArea');
        this.dataAreaActive = true;
        this.titleActive = this.subTitleActive = this.headerActive = this.footerActive = false;
    };
    /**
     * @return {?}
     */
    StyleEditorComponent.prototype.handleFooterBtn = /**
     * @return {?}
     */
    function () {
        this.updateStyleOption();
        this.initStyleOption('Footer');
        this.footerActive = true;
        this.titleActive = this.subTitleActive = this.headerActive = this.dataAreaActive = false;
    };
    /**
     * @return {?}
     */
    StyleEditorComponent.prototype.handleWeight = /**
     * @return {?}
     */
    function () {
        this.isWeight = !this.isWeight;
        if (this.isWeight) {
            this.fontWeight = 'bold';
        }
        else {
            this.fontWeight = 'normal';
        }
    };
    /**
     * @return {?}
     */
    StyleEditorComponent.prototype.handleStyle = /**
     * @return {?}
     */
    function () {
        this.isStyle = !this.isStyle;
        if (this.isStyle) {
            this.fontStyle = 'italic';
        }
        else {
            this.fontStyle = 'normal';
        }
    };
    /**
     * @return {?}
     */
    StyleEditorComponent.prototype.handleUnderLine = /**
     * @return {?}
     */
    function () {
        this.isUnderLine = !this.isUnderLine;
        if (this.isUnderLine) {
            this.fontUnderLine = 'underline';
        }
        else {
            this.fontUnderLine = 'none';
        }
    };
    /**
     * @return {?}
     */
    StyleEditorComponent.prototype.getStyleOption = /**
     * @return {?}
     */
    function () {
        this.updateStyleOption();
        return this.styleOption;
    };
    /**
     * @private
     * @param {?} key
     * @return {?}
     */
    StyleEditorComponent.prototype.initStyleOption = /**
     * @private
     * @param {?} key
     * @return {?}
     */
    function (key) {
        if (this.styleOption && this.styleOption[key]) {
            this.rowHeight = this.styleOption[key].rowHeight - 0;
            this.fontFamily = this.styleOption[key].fontFamily - 0;
            this.fontSize = this.styleOption[key].size;
            this.fontWeight = this.styleOption[key].fontWeight === 0 ? 'normal' : 'bold';
            this.fontStyle = this.styleOption[key].fontStyle === 0 ? 'normal' : 'italic';
            this.fontUnderLine = this.styleOption[key].textDecoration === 0 ? 'none' : 'underline';
            this.fontColor = this.styleOption[key].fontColor;
        }
    };
    /**
     * @private
     * @return {?}
     */
    StyleEditorComponent.prototype.updateStyleOption = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var key = '';
        if (this.titleActive) {
            key = 'Title';
        }
        else if (this.subTitleActive) {
            key = 'SubTitle';
        }
        else if (this.headerActive) {
            key = 'Header';
        }
        else if (this.dataAreaActive) {
            key = 'DataArea';
        }
        else {
            key = 'Footer';
        }
        if (this.styleOption && this.styleOption[key]) {
            this.styleOption[key].rowHeight = this.rowHeight;
            this.styleOption[key].fontFamily = this.fontFamily;
            this.styleOption[key].size = this.fontSize - 0;
            this.styleOption[key].fontWeight = this.fontWeight === 'normal' ? 0 : 1;
            this.styleOption[key].fontStyle = this.fontStyle === 'normal' ? 0 : 1;
            this.styleOption[key].textDecoration = this.fontUnderLine === 'none' ? 0 : 1;
            this.styleOption[key].fontColor = this.fontColor;
        }
    };
    /**
     * @param {?} key
     * @return {?}
     */
    StyleEditorComponent.prototype.getFontFamily = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        switch (key - 0) {
            case 0:
                return 'SimSun';
            case 1:
                return 'NSimSun';
            case 2:
                return 'FangSong';
            case 3:
                return 'KaiTi';
            case 4:
                return 'SimHei';
            case 5:
                return 'ARIAL';
            case 6:
                return 'ROMAN';
            case 7:
                return '\'TIMES NEW ROMAN\'';
            case 8:
                return 'VERDANA';
            case 9:
                return '\'Microsoft YaHei\'';
        }
    };
    StyleEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'style-editor',
                    template: "<div class=\"d-flex flex-column flex-fill h-100\">\r\n  <div class=\"f-utils-fill-flex-column\">\r\n    <div class=\"f-utils-fill-flex-column\">\r\n      <div class=\"mx-2 row no-gutters\">\r\n        <div class=\"col-9 col-md-9 col-lg-9\">\r\n          <div class=\"farris-group-wrap\">\r\n            <div class=\"form-group farris-form-group d-flex flex-column h-100\">\r\n              <label class=\"col-form-label\" for=\"fontBanner\" style=\"width: 5rem; min-width: 5rem;\">\r\n                <span class=\"farris-label-text\">{{localizeService.getValue('ideCmp.schemaManager.styleOption.column')}}:</span>\r\n              </label>\r\n              <div class=\"farris-input-wrap\">\r\n                <button class=\"btn mr-1\" [class.btn-primary]=\"titleActive\" [class.btn-secondary]=\"!titleActive\" (click)=\"handleTitleBtn()\"\r\n                  id=\"button-edit\">\r\n                  {{localizeService.getValue('ideCmp.schemaManager.styleOption.styleTitle')}}</button>\r\n                <button class=\"btn mr-1\" [class.btn-primary]=\"subTitleActive\" [class.btn-secondary]=\"!subTitleActive\"\r\n                  (click)=\"handleSubTitleBtn()\" id=\"button-edit\">\r\n                  {{localizeService.getValue('ideCmp.schemaManager.styleOption.styleSubTitle')}}</button>\r\n                <button class=\"btn mr-1\" [class.btn-primary]=\"headerActive\" [class.btn-secondary]=\"!headerActive\"\r\n                  (click)=\"handleHeaderBtn()\" id=\"button-edit\">\r\n                  {{localizeService.getValue('ideCmp.schemaManager.styleOption.styleHeader')}}</button>\r\n                <button class=\"btn mr-1\" [class.btn-primary]=\"dataAreaActive\" [class.btn-secondary]=\"!dataAreaActive\"\r\n                  (click)=\"handleDataAreaBtn()\" id=\"button-edit\">\r\n                  {{localizeService.getValue('ideCmp.schemaManager.styleOption.styleBody')}}</button>\r\n                <button class=\"btn\" [class.btn-primary]=\"footerActive\" [class.btn-secondary]=\"!footerActive\" (click)=\"handleFooterBtn()\"\r\n                  id=\"button-edit\">\r\n                  {{localizeService.getValue('ideCmp.schemaManager.styleOption.styleFooter')}}</button>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n        <div class=\"col-3 col-md-3 col-lg-3\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\" for=\"rowHeight\">\r\n              <span class=\"farris-label-text\">{{localizeService.getValue('ideCmp.schemaManager.styleOption.rowHeight')}}:</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <farris-number-spinner id=\"rowHeight\" [(ngModel)]=\"rowHeight\" [min]=\"1\">\r\n              </farris-number-spinner>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <div class=\"mx-2 row no-gutters\">\r\n        <div class=\"col-4 col-md-4 col-lg-4\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\" for=\"rowHeight\">\r\n              <span class=\"farris-label-text\">{{localizeService.getValue('ideCmp.schemaManager.styleOption.textStyle')}}:</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n                <farris-combo-list [editable]=\"true\" [autoWidth]=\"false\" [(ngModel)]=\"fontFamily\" [data]=\"fontFamilyData\"\r\n                  [textField]=\"'name'\" [idField]=\"'value'\"></farris-combo-list>\r\n            </div>\r\n          </div>\r\n        </div>\r\n        <div class=\"col-4 col-md-4 col-lg-4\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\" for=\"rowHeight\">\r\n              <span class=\"farris-label-text\">&nbsp;</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n                <farris-combo-list style=\"width: 20%\" [editable]=\"true\" [autoWidth]=\"false\" [(ngModel)]=\"fontSize\"\r\n                  [data]=\"fontSizeData\" [textField]=\"'name'\" [idField]=\"'value'\"></farris-combo-list>\r\n            </div>\r\n          </div>\r\n        </div>\r\n        <div class=\"col-4 col-md-4 col-lg-4\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\" for=\"rowHeight\">\r\n              <span class=\"farris-label-text\">&nbsp;</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <button class=\"k-button k-button-icon k-flat\" id=\"button-search\" (click)=\"handleWeight()\">\r\n                <span class=\"k-icon k-i-bold\" role=\"presentation\"></span>\r\n              </button>\r\n              <button class=\"k-button k-button-icon k-flat\" id=\"button-search\" (click)=\"handleStyle()\">\r\n                <span class=\"k-icon k-i-italic\" role=\"presentation\"></span>\r\n              </button>\r\n              <button class=\"k-button k-button-icon k-flat\" id=\"button-search\" (click)=\"handleUnderLine()\">\r\n                <span class=\"k-icon k-i-underline\" role=\"presentation\"></span>\r\n              </button>\r\n              <button class=\"k-button k-button-icon k-flat\" id=\"button-search\" [(colorPicker)]=\"fontColor\"\r\n                [cpOKButton]=\"true\" [cpCancelButton]=\"true\">\r\n                <span [style.border-bottom-color]=\"fontColor\" class=\"k-icon k-i-foreground-color\"\r\n                  style=\"border-bottom-width: 2px; border-bottom-style: solid; margin-top: -2px;\"\r\n                  role=\"presentation\"></span>\r\n              </button>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <div class=\"flex-fill position-relative\">\r\n        <div class=\"d-flex flex-row\"\r\n          style=\"margin-bottom: .375rem; position: absolute; top: 0; bottom: 0; left: 10px; right: 10px;\">\r\n          <fieldSet style=\"border: 1px solid #aaa !important; width: 100% !important;\">\r\n            <legend class=\"ml-2 p-0\">{{localizeService.getValue('ideCmp.schemaManager.styleOption.preview')}}</legend>\r\n            <div class=\"d-flex flex-wrap mb-2\" style=\"position: absolute; top: 0; left: 0; right: 0; bottom: 0;\">\r\n              <span class=\"preview-span\" [style.text-decoration]=\"fontUnderLine\" [style.font-style]=\"fontStyle\"\r\n                [style.font-weight]=\"fontWeight\" [style.font-family]=\"getFontFamily(fontFamily)\"\r\n                [style.color]=\"fontColor\" [style.font-size.pt]=\"fontSize\">AaBbYyZz</span>\r\n            </div>\r\n          </fieldSet>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</div>\r\n",
                    styles: [".preview-span{width:100%!important;top:50%;left:50%;-webkit-transform:translate(-50%,-50%);transform:translate(-50%,-50%);text-align:center;position:absolute}"]
                }] }
    ];
    /** @nocollapse */
    StyleEditorComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    StyleEditorComponent.propDecorators = {
        currSchema: [{ type: Input }]
    };
    return StyleEditorComponent;
}());
export { StyleEditorComponent };
if (false) {
    /** @type {?} */
    StyleEditorComponent.prototype.currSchema;
    /** @type {?} */
    StyleEditorComponent.prototype.styleOption;
    /** @type {?} */
    StyleEditorComponent.prototype.rowHeight;
    /** @type {?} */
    StyleEditorComponent.prototype.fontFamilyData;
    /** @type {?} */
    StyleEditorComponent.prototype.fontFamily;
    /** @type {?} */
    StyleEditorComponent.prototype.fontSizeData;
    /** @type {?} */
    StyleEditorComponent.prototype.fontSize;
    /** @type {?} */
    StyleEditorComponent.prototype.fontColor;
    /**
     * @type {?}
     * @private
     */
    StyleEditorComponent.prototype.isWeight;
    /**
     * @type {?}
     * @private
     */
    StyleEditorComponent.prototype.isStyle;
    /**
     * @type {?}
     * @private
     */
    StyleEditorComponent.prototype.isUnderLine;
    /** @type {?} */
    StyleEditorComponent.prototype.fontWeight;
    /** @type {?} */
    StyleEditorComponent.prototype.fontStyle;
    /** @type {?} */
    StyleEditorComponent.prototype.fontUnderLine;
    /** @type {?} */
    StyleEditorComponent.prototype.titleActive;
    /** @type {?} */
    StyleEditorComponent.prototype.subTitleActive;
    /** @type {?} */
    StyleEditorComponent.prototype.headerActive;
    /** @type {?} */
    StyleEditorComponent.prototype.dataAreaActive;
    /** @type {?} */
    StyleEditorComponent.prototype.footerActive;
    /** @type {?} */
    StyleEditorComponent.prototype.localizeService;
    /**
     * @type {?}
     * @private
     */
    StyleEditorComponent.prototype.injector;
}
