// @progress/kendo-angular-layout/dist/npm/common/draggable.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var kendo_draggable_1 = require("@telerik/kendo-draggable");
/**
 * @hidden
 */
var DraggableDirective = /** @class */ (function () {
    function DraggableDirective(zone, element) {
        var _this = this;
        this.zone = zone;
        this.kendo = {
            drag: new core_1.EventEmitter(),
            press: new core_1.EventEmitter(),
            release: new core_1.EventEmitter()
        };
        if (typeof document !== 'undefined') {
            this.draggable = new kendo_draggable_1.default({
                drag: function (e) { return _this.kendo.drag.next(e); },
                press: function (e) { return _this.kendo.press.next(e); },
                release: function (e) { return _this.kendo.release.next(e); }
            });
            this.zone.runOutsideAngular(function () { return _this.draggable.bindTo(element.nativeElement); });
        }
    }
    DraggableDirective.prototype.ngOnDestroy = function () {
        if (typeof document !== 'undefined') {
            this.draggable.destroy();
        }
    };
    DraggableDirective.decorators = [
        { type: core_1.Directive, args: [{
                    outputs: ['kendo.press', 'kendo.drag', 'kendo.release'],
                    selector: '[kendoLayoutDraggable]'
                },] },
    ];
    /** @nocollapse */
    DraggableDirective.ctorParameters = function () { return [
        { type: core_1.NgZone, },
        { type: core_1.ElementRef, },
    ]; };
    return DraggableDirective;
}());
exports.DraggableDirective = DraggableDirective;
