// @farris/command-services/esm2015/lib/work-flow/work-flow-message.service.js
import { Injectable } from "@angular/core";
export class WorkFlowMessageService {
    constructor() {
        this.messageHandler = (callback) => {
            return (event) => {
                var message = event.data;
                if (message) {
                    message.sender = event.source;
                }
                if (typeof callback === 'function') {
                    callback(message);
                }
            };
        };
    }
    /**
     * 监听工作流消息
     * @param callback 回调
     */
    listen(callback) {
        this.handler = this.messageHandler(callback);
        window.addEventListener('message', this.handler);
    }
    destroy() {
        if (this.handler) {
            window.removeEventListener('message', this.handler);
        }
    }
    /**
     * 给工作流表单发送消息
     * @param message
     */
    send(message) {
        if (!message) {
            return;
        }
        let win = window;
        if (message.target) {
            win = message.target || window;
            delete message.target;
        }
        if (win) {
            win.postMessage(message, '*');
        }
    }
}
WorkFlowMessageService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
WorkFlowMessageService.ctorParameters = () => [];
