// @farris/command-services/esm2015/lib/work-flow/work-flow-message.js
import { Injectable, Injector } from "@angular/core";
import { WorkFlowMessageService } from "./work-flow-message.service";
/**
 * 工作流消息
 * @warning 禁止使用表单上下文，所有上下文必须从外部传入
 */
export class WorkFlowMessage {
    constructor(injector, workFlowMessageService) {
        this.injector = injector;
        this.workFlowMessageService = workFlowMessageService;
        this.listeners = new Map();
    }
    setup() {
        if (!this.workFlowMessageService) {
            return;
        }
        this.workFlowMessageService.listen((message) => {
            if (!this.listeners || this.listeners.size < 1) {
                return;
            }
            this.listeners.forEach((callback, key) => {
                if (typeof callback === 'function') {
                    callback(message);
                }
            });
        });
    }
    addEventListener(callback) {
        const id = `${new Date().valueOf().toString(16)}-${Math.ceil(Math.random() * 100)}`;
        this.listeners.set(id, callback);
        return id;
    }
    removeEventListener(eventId) {
        if (!this.listeners || this.listeners.size < 1) {
            return;
        }
        this.listeners.delete(eventId);
        eventId = null;
    }
}
WorkFlowMessage.decorators = [
    { type: Injectable }
];
/** @nocollapse */
WorkFlowMessage.ctorParameters = () => [
    { type: Injector },
    { type: WorkFlowMessageService }
];
