// @farris/mobile-command-services/esm5/lib/business-attachment-services/business-attachment.util.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/business-attachment-services/business-attachment.util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetailType } from '@farris/mobile-bef';
/**
 * 附件信息处理工具类
 */
var /**
 * 附件信息处理工具类
 */
AttachmentUtil = /** @class */ (function () {
    function AttachmentUtil() {
    }
    /**
     * 转换为附件信息数组
     */
    /**
     * 转换为附件信息数组
     * @param {?} fileInfos
     * @return {?}
     */
    AttachmentUtil.convertToAttachmentInfos = /**
     * 转换为附件信息数组
     * @param {?} fileInfos
     * @return {?}
     */
    function (fileInfos) {
        var _this = this;
        if (!fileInfos) {
            return [];
        }
        /** @type {?} */
        var attachmentInfos = fileInfos.map((/**
         * @param {?} fileInfo
         * @return {?}
         */
        function (fileInfo) {
            return _this.convertToAttachmentInfo(fileInfo);
        }));
        return attachmentInfos;
    };
    /**
     * 转换为附件信息
     */
    /**
     * 转换为附件信息
     * @param {?} fileInfo
     * @return {?}
     */
    AttachmentUtil.convertToAttachmentInfo = /**
     * 转换为附件信息
     * @param {?} fileInfo
     * @return {?}
     */
    function (fileInfo) {
        /** @type {?} */
        var attachmentInfo = {
            attachmentId: fileInfo.metadataId,
            fileName: fileInfo.fileName
        };
        return attachmentInfo;
    };
    /**
     * 获取附件列表中的第一个附件
     */
    /**
     * 获取附件列表中的第一个附件
     * @param {?} attachmentInfos
     * @return {?}
     */
    AttachmentUtil.getFirstAttachmentInfo = /**
     * 获取附件列表中的第一个附件
     * @param {?} attachmentInfos
     * @return {?}
     */
    function (attachmentInfos) {
        if (!attachmentInfos || attachmentInfos.length === 0) {
            return;
        }
        /** @type {?} */
        var firstBusinessAttachmentInfo = attachmentInfos[0];
        return firstBusinessAttachmentInfo;
    };
    /**
     * 提取附件id数组
     */
    /**
     * 提取附件id数组
     * @param {?} attachmentInfos
     * @return {?}
     */
    AttachmentUtil.peekAttachmentIds = /**
     * 提取附件id数组
     * @param {?} attachmentInfos
     * @return {?}
     */
    function (attachmentInfos) {
        if (!attachmentInfos) {
            attachmentInfos = [];
        }
        /** @type {?} */
        var attachmentIds = attachmentInfos.map((/**
         * @param {?} attachmentInfo
         * @return {?}
         */
        function (attachmentInfo) {
            return attachmentInfo.attachmentId;
        }));
        return attachmentIds;
    };
    /**
     * 批量获取变更
     */
    /**
     * 批量获取变更
     * @param {?} oldListData
     * @param {?} newListData
     * @return {?}
     */
    AttachmentUtil.getChangeDetails = /**
     * 批量获取变更
     * @param {?} oldListData
     * @param {?} newListData
     * @return {?}
     */
    function (oldListData, newListData) {
        var _this = this;
        /** @type {?} */
        var changeDetails = [];
        newListData.forEach((/**
         * @param {?} newData
         * @return {?}
         */
        function (newData) {
            /** @type {?} */
            var oldData = oldListData.find((/**
             * @param {?} oldData
             * @return {?}
             */
            function (oldData) {
                return oldData['id'] === newData['id'];
            }));
            if (!oldData) {
                oldData = {};
            }
            /** @type {?} */
            var changeDetail = _this.getChangeDetail(oldData, newData);
            if (changeDetail) {
                changeDetails.push(changeDetail);
            }
        }));
        return changeDetails;
    };
    /**
     * 获取变更集
     */
    /**
     * 获取变更集
     * @param {?} oldData
     * @param {?} newData
     * @return {?}
     */
    AttachmentUtil.getChangeDetail = /**
     * 获取变更集
     * @param {?} oldData
     * @param {?} newData
     * @return {?}
     */
    function (oldData, newData) {
        /** @type {?} */
        var changeDetail = {
            ChangeType: ChangeDetailType.Modify,
            ChangeInfo: {
                DataId: newData['id']
            }
        };
        /** @type {?} */
        var isChanged = false;
        Object.keys(newData).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        function (propName) {
            // 移除不必要的属性的比较
            if (propName === 'id' || propName === 'nodeCode' || propName === 'childs') {
                return;
            }
            /** @type {?} */
            var oldValue = oldData[propName];
            /** @type {?} */
            var newValue = newData[propName];
            if (oldValue === undefined && typeof newValue === 'object') {
                return;
            }
            if (JSON.stringify(oldValue) !== JSON.stringify(newValue)) {
                changeDetail.ChangeInfo[propName] = newValue;
                isChanged = true;
            }
        }));
        return isChanged ? changeDetail : null;
    };
    return AttachmentUtil;
}());
export { AttachmentUtil };
