// @farris/mobile-command-services/esm5/lib/work-flow-service/submit/process-selector-vue/process-selector.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/work-flow-service/submit/process-selector-vue/process-selector.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Subject } from 'rxjs';
import { StartProcessPayload, FlowNodeInfo } from '../../entity/start-process-payload';
import { SubmitApprove } from '@farris/mobile-business-ui';
var ProcessSelectorService = /** @class */ (function () {
    function ProcessSelectorService(taskService) {
        this.taskService = taskService;
        //候选下一环节
        this.processPlaceholder = '选择流程'; //选择流程placeholder
        //保存模拟流程后返回的内容，选择节点时使用
        this.assignInfos = [];
        //选择的下一环节ID，后续环节id的拼接
        this.selectedValues = ['clear'];
        this.payload = new StartProcessPayload; //保存表单上传来的payload
        //保存表单上传来的payload
        this.subject = new Subject();
        this.processReadOnly = false; //选择流程框是否只读
        //选择流程框是否只读
        this.nodeReadOnly = true; //选择节点下拉框是否只读
    }
    //弹框实例
    /**
     * @param {?} param
     * @return {?}
     */
    ProcessSelectorService.prototype.setParam = 
    //弹框实例
    /**
     * @param {?} param
     * @return {?}
     */
    function (param) {
        this.payload.id = param.id;
        this.payload.processDefinitionId = param.processDefinitionId;
        this.payload.processDefinitionKey = param.processDefinitionKey;
        this.payload.bizDefKey = param.bizDefKey;
        this.payload.dataModelId = param.dataModelId;
        this.payload.dataId = param.dataId;
        this.payload.name = param.name;
        this.payload.startUserId = param.startUserId;
        this.payload.nextFlowNodes = [];
        this.startProcessSimulate(this.payload);
    };
    /**
     * @param {?} payload
     * @return {?}
     */
    ProcessSelectorService.prototype.startProcessSimulate = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        var _this = this;
        this.taskService.startProcessSimulate(payload).subscribe((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            _this.destProcesses = response.procDefs.map((/**
             * @param {?} process
             * @return {?}
             */
            function (process) { return ({ id: process.procDefId, name: process.procDefName }); }));
            if (_this.destProcesses.length === 1) {
                _this.processReadOnly = true;
                _this.selectedProcessId = response.procDefs[0].procDefId;
                _this.vm = ((/** @type {?} */ (SubmitApprove))).confirm({
                    process: _this.selectedProcessId,
                    processData: _this.destProcesses,
                    processOc: _this.selectProcessChange.bind(_this),
                    processReadFlag: _this.processReadOnly,
                    processHolder: _this.processPlaceholder,
                    nextNodeReadFlag: _this.nodeReadOnly,
                    submit: _this.submit.bind(_this),
                    cancel: _this.back.bind(_this)
                });
                _this.response = response;
                _this.setNextNode(response);
            }
            else {
                _this.vm = ((/** @type {?} */ (SubmitApprove))).confirm({
                    processData: _this.destProcesses,
                    processOc: _this.selectProcessChange.bind(_this),
                    processReadFlag: _this.processReadOnly,
                    processHolder: _this.processPlaceholder,
                    nextNodeReadFlag: _this.nodeReadOnly,
                    submit: _this.submit.bind(_this),
                    cancel: _this.back.bind(_this)
                });
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.subject.next({ type: 'error', message: error && error.response && error.response.data && error.response.data.Message || '提交失败' });
        }));
    };
    /**
   * 选择某个流程后
   * @param data
   */
    /**
     * 选择某个流程后
     * @param {?} data
     * @return {?}
     */
    ProcessSelectorService.prototype.selectProcessChange = /**
     * 选择某个流程后
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        this.clearHelp();
        this.clearNode();
        this.selectedProcessId = data;
        /** @type {?} */
        var payload = new StartProcessPayload();
        payload.processDefinitionId = this.selectedProcessId;
        payload.dataId = this.payload.dataId;
        // const load = this.loadingService.show();
        this.taskService.startProcessSimulate(payload).subscribe((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            // load.close();
            _this.response = response;
            _this.setNextNode(response);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            //load.close();
            _this.subject.next({ type: 'error', message: error && error.response && error.response.data && error.response.data.Message || '该流程存在问题' });
        }));
    };
    /**
     * @param {?} data
     * @return {?}
     */
    ProcessSelectorService.prototype.getsimulationPathsIndexData = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        return this.destNodes.findIndex((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.id === data; }));
    };
    /**
   * 选择节点后事件
   */
    /**
     * 选择节点后事件
     * @param {?} data
     * @return {?}
     */
    ProcessSelectorService.prototype.selectNodeChange = /**
     * 选择节点后事件
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        this.clearHelp();
        this.clearAssignInfos();
        this.selectedNodeId = data;
        this.vm.open({
            nextNode: this.selectedNodeId,
        });
        //const nodeNames = data.data.name.split('，');
        /** @type {?} */
        var index = this.getsimulationPathsIndexData(data);
        if (this.response.simulationPaths[index].nodeInfoList.every((/**
         * @param {?} node
         * @return {?}
         */
        function (node) { return node.resolved === true; }))) { //已预解析参与者
            this.setAssignInfo(this.response.simulationPaths[index]);
        }
        else { //未预解析参与者
            //未预解析参与者
            /** @type {?} */
            var payload_1 = new StartProcessPayload();
            payload_1.processDefinitionId = this.selectedProcessId;
            payload_1.dataId = this.payload.dataId;
            payload_1.nextFlowNodes = [];
            this.selectedNodeId.split('，').forEach((/**
             * @param {?} nodeId
             * @param {?} index
             * @return {?}
             */
            function (nodeId, index) {
                /** @type {?} */
                var nextFlowNode = new FlowNodeInfo();
                nextFlowNode.activityDefinitionId = nodeId;
                //nextFlowNode.activityDefinitionName = nodeNames[index];
                payload_1.nextFlowNodes.push(nextFlowNode);
            }));
            //const load = this.loadingService.show();
            this.taskService.startProcessSimulate(payload_1).subscribe((/**
             * @param {?} response
             * @return {?}
             */
            function (response) {
                //load.close();
                _this.setAssignInfo(response.simulationPaths[0]); //选择的下一环节应该在同一条路径上，故取simulationPaths[0]
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) {
                //load.close();
                _this.subject.next({ type: 'error', message: error && error.response && error.response.data && error.response.data.Message || '该节点存在问题' });
            }));
        }
    };
    /**
   * 设置下一环节
   */
    /**
     * 设置下一环节
     * @private
     * @param {?} response
     * @return {?}
     */
    ProcessSelectorService.prototype.setNextNode = /**
     * 设置下一环节
     * @private
     * @param {?} response
     * @return {?}
     */
    function (response) {
        this.destNodes = response.simulationPaths && response.simulationPaths.map((/**
         * @param {?} path
         * @return {?}
         */
        function (path) { return ({
            id: path.nodeInfoList.map((/**
             * @param {?} node
             * @return {?}
             */
            function (node) { return node.activityDefinitionId; })).join('，'),
            name: path.nodeInfoList.map((/**
             * @param {?} node
             * @return {?}
             */
            function (node) { return node.activityDefinitionName; })).join('，')
        }); }));
        //处理默认分支，无默认分支返回-1
        /** @type {?} */
        var defaultIndex = response.simulationPaths.findIndex((/**
         * @param {?} a
         * @return {?}
         */
        function (a) { return a.default; }));
        if (defaultIndex > -1) {
            this.destNodes.length === 1 ? (this.nodeReadOnly = true) : (this.nodeReadOnly = false);
            this.selectedNodeId = this.destNodes[defaultIndex].id;
            this.setAssignInfo(response.simulationPaths[defaultIndex]);
        }
        else {
            this.selectedNodeId = '';
            this.nodeReadOnly = false;
            this.nodePlaceholder = '请选择';
        }
        this.vm.open({
            nextNodeOc: this.selectNodeChange.bind(this),
            nextNode: this.selectedNodeId,
            nextNodeData: this.destNodes,
            nextNodeReadFlag: this.nodeReadOnly,
            nextNodeHolder: this.nodePlaceholder
        });
    };
    /**
     * 设置当前指派信息帮助参数
     */
    /**
     * 设置当前指派信息帮助参数
     * @private
     * @param {?} simulationPaths
     * @return {?}
     */
    ProcessSelectorService.prototype.setAssignInfo = /**
     * 设置当前指派信息帮助参数
     * @private
     * @param {?} simulationPaths
     * @return {?}
     */
    function (simulationPaths) {
        var _this = this;
        this.assignInfos = [];
        this.selectedNodeId.split('，').forEach((/**
         * @param {?} nodeId
         * @return {?}
         */
        function (nodeId) {
            /** @type {?} */
            var node = simulationPaths.nodeInfoList.find((/**
             * @param {?} node
             * @return {?}
             */
            function (node) { return node.activityDefinitionId === nodeId; }));
            if (node) {
                /** @type {?} */
                var selectedAssignees = node.selectedAssignees;
                /** @type {?} */
                var assigneeReadFlag = false;
                /** @type {?} */
                var assignee_1 = '';
                /** @type {?} */
                var assigneeHolder = '请选择人员';
                if (selectedAssignees && selectedAssignees.length >= 1) {
                    assigneeReadFlag = true;
                    assigneeHolder = "";
                    selectedAssignees.forEach((/**
                     * @param {?} user
                     * @return {?}
                     */
                    function (user) {
                        assignee_1 = assignee_1 + user.name + ', ';
                    }));
                    assignee_1 = assignee_1.slice(0, assignee_1.length - 2);
                }
                if (node.anonymous) {
                    assignee_1 = node.anonymousText === '' ? '' : node.anonymousText && node.anonymousText;
                }
                _this.assignInfos.push({
                    label: (_this.selectedNodeId && _this.selectedNodeId.split('，').length > 1 ? node.activityDefinitionName : '') + "\u529E\u7406\u4EBA\u5458",
                    selectedAssignees: selectedAssignees,
                    assigneeReadFlag: assigneeReadFlag,
                    assignee: assignee_1,
                    assigneeHolder: assigneeHolder,
                    assigneeInfo: tslib_1.__assign({}, node.assigneeInfo, { assigneeUsers: { items: (node.assigneeInfo && node.assigneeInfo.assigneeUsers) || [] } }),
                    assignees: node.selectedAssignees.map((/**
                     * @param {?} user
                     * @return {?}
                     */
                    function (user) { return { id: user.id, name: user.name }; })),
                    activityName: node.activityDefinitionName,
                    dependency: node.dependency,
                    anonymous: node.anonymous,
                    anonymousText: node.anonymousText
                });
            }
            else {
                _this.assignInfos.push({
                    selectedAssignees: [],
                    assigneeInfo: null,
                    assignees: [],
                    activityName: '',
                    dependency: '',
                    anonymous: node.anonymous,
                    anonymousText: node.anonymousText
                });
            }
        }));
        this.vm.open({
            assigneeOc: this.assigneeOc.bind(this),
            assignInfos: this.assignInfos
        });
    };
    // 办理人员选择事件
    // 办理人员选择事件
    /**
     * @param {?} data
     * @param {?} index
     * @return {?}
     */
    ProcessSelectorService.prototype.assigneeOc = 
    // 办理人员选择事件
    /**
     * @param {?} data
     * @param {?} index
     * @return {?}
     */
    function (data, index) {
        this.assignInfos[index].assignees = data.items;
        this.selectedValues[index] = data.value;
        this.vm.open({
            selectedValues: this.selectedValues
        });
    };
    /**
     * @return {?}
     */
    ProcessSelectorService.prototype.clearHelp = /**
     * @return {?}
     */
    function () {
        this.vm.open({
            selectedValues: ['clear']
        });
    };
    // 清空节点
    // 清空节点
    /**
     * @return {?}
     */
    ProcessSelectorService.prototype.clearNode = 
    // 清空节点
    /**
     * @return {?}
     */
    function () {
        this.selectedNodeId = '',
            this.destNodes = [],
            this.nodeReadOnly = true,
            this.nodePlaceholder = '',
            this.assignInfos = [];
        this.vm.open({
            nextNode: this.selectedNodeId,
            nextNodeData: this.destNodes,
            nextNodeReadFlag: this.nodeReadOnly,
            nodePlaceholder: this.nodePlaceholder,
            assignInfos: this.assignInfos
        });
    };
    // 清空办理人员
    // 清空办理人员
    /**
     * @return {?}
     */
    ProcessSelectorService.prototype.clearAssignInfos = 
    // 清空办理人员
    /**
     * @return {?}
     */
    function () {
        this.assignInfos = [];
        this.vm.open({
            assignInfos: this.assignInfos
        });
    };
    /**
     * 取消页面弹框
     */
    /**
     * 取消页面弹框
     * @return {?}
     */
    ProcessSelectorService.prototype.back = /**
     * 取消页面弹框
     * @return {?}
     */
    function () {
        this.clearProcess();
        this.subject.next({ type: 'success', message: '' });
    };
    /**
     * @return {?}
     */
    ProcessSelectorService.prototype.submit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.selectedProcessId) {
            this.subject.next({ type: 'error', message: '请选流程' });
            return;
        }
        if (!this.selectedNodeId) {
            this.subject.next({ type: 'error', message: '请选节点' });
            return;
        }
        if (this.assignInfos.filter((/**
         * @param {?} info
         * @return {?}
         */
        function (info) { return ((info.assigneeInfo && info.assigneeInfo.assigneeUsers.items.length) || info.selectedAssignees.length) && !info.assignees.length; })).length) {
            this.subject.next({ type: 'error', message: '请选办理人员' });
            return;
        }
        /** @type {?} */
        var nodes = this.selectedNodeId.split('，');
        for (var i = 0; i < nodes.length; i++) {
            /** @type {?} */
            var nextFlowNode = {
                activityDefinitionId: nodes[i],
                activityDefinitionName: this.assignInfos[i].activityName,
                dependency: this.assignInfos[i].dependency,
                selectedAssignees: this.assignInfos[i].assigneeInfo ? this.assignInfos[i].assignees : [],
                assigneeInfo: null,
                resolved: false,
                anonymous: false,
                anonymousText: ''
            };
            this.payload.nextFlowNodes.push(nextFlowNode);
        }
        this.payload.processDefinitionId = this.selectedProcessId;
        this.taskService.startProcess(this.payload).subscribe((/**
         * @return {?}
         */
        function () {
            _this.clearProcess();
            _this.subject.next({ type: 'success', message: '提交成功' });
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.subject.next({ type: 'error', message: error && error.response && error.response.data && error.response.data.Message || '提交失败' });
        }));
    };
    /**
     * @return {?}
     */
    ProcessSelectorService.prototype.getSubject = /**
     * @return {?}
     */
    function () {
        return this.subject.asObservable();
    };
    /**
     * @return {?}
     */
    ProcessSelectorService.prototype.clearProcess = /**
     * @return {?}
     */
    function () {
        this.vm.close();
        this.vm = null;
        this.selectedProcessId = '';
        this.selectedNodeId = '';
        this.processReadOnly = false;
        this.nodeReadOnly = true;
    };
    return ProcessSelectorService;
}());
export { ProcessSelectorService };
if (false) {
    /** @type {?} */
    ProcessSelectorService.prototype.destProcesses;
    /** @type {?} */
    ProcessSelectorService.prototype.destNodes;
    /** @type {?} */
    ProcessSelectorService.prototype.processPlaceholder;
    /** @type {?} */
    ProcessSelectorService.prototype.nodePlaceholder;
    /** @type {?} */
    ProcessSelectorService.prototype.response;
    /** @type {?} */
    ProcessSelectorService.prototype.assignInfos;
    /** @type {?} */
    ProcessSelectorService.prototype.selectedProcessId;
    /** @type {?} */
    ProcessSelectorService.prototype.selectedNodeId;
    /** @type {?} */
    ProcessSelectorService.prototype.selectedValues;
    /** @type {?} */
    ProcessSelectorService.prototype.payload;
    /** @type {?} */
    ProcessSelectorService.prototype.subject;
    /** @type {?} */
    ProcessSelectorService.prototype.processReadOnly;
    /** @type {?} */
    ProcessSelectorService.prototype.nodeReadOnly;
    /** @type {?} */
    ProcessSelectorService.prototype.vm;
    /**
     * @type {?}
     * @private
     */
    ProcessSelectorService.prototype.taskService;
}
