// @ecp-caf/common-structure/esm5/lib/convertor/parameter/parameter-convertor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DataTypeConvertUtil } from '../utils/data-type-convert-util';
import { ExtensionConvertor } from '../extension/extension-convertor';
var ParameterConvertor = /** @class */ (function () {
    function ParameterConvertor() {
    }
    /**
     * @param {?} obj
     * @param {?} parameter
     * @param {?} ctxt
     * @return {?}
     */
    ParameterConvertor.prototype.convertToObject = /**
     * @param {?} obj
     * @param {?} parameter
     * @param {?} ctxt
     * @return {?}
     */
    function (obj, parameter, ctxt) {
        if (!parameter) {
            return;
        }
        obj['kind'] = parameter.kind;
        obj['code'] = parameter.code;
        obj['name'] = parameter.name;
        obj['description'] = parameter.description;
        obj['required'] = parameter.required;
        obj['defaultValue'] = parameter.defaultValue;
        obj['parameterType'] = null;
        if (parameter.parameterType) {
            /** @type {?} */
            var kind = parameter.parameterType.getKind();
            if (!kind) {
                throw new Error('参数序列化时数据类型的TypeKind为空！');
            }
            obj['parameterType'] = DataTypeConvertUtil.convertToObject(parameter.parameterType, ctxt);
        }
        obj['constraint'] = null;
        if (parameter.constraint) {
            /** @type {?} */
            var kind = parameter.constraint.getKind();
            if (!kind) {
                throw new Error('参数序列化时参数约束的ConstraintKind为空！');
            }
            obj['constraint'] = ctxt.getConstraintSerializerFactory().get(kind).serialize(parameter.constraint, ctxt);
        }
        if (parameter.extension) {
            /** @type {?} */
            var extensionConvertor = new ExtensionConvertor();
            extensionConvertor.addToJObj(parameter, obj);
        }
    };
    /**
     * @param {?} parameter
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    ParameterConvertor.prototype.initFromObject = /**
     * @param {?} parameter
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    function (parameter, obj, ctxt) {
        if (!obj) {
            return;
        }
        parameter.kind = obj['kind'];
        parameter.code = obj['code'];
        parameter.name = obj['name'];
        parameter.defaultValue = obj['defaultValue'];
        parameter.description = obj['description'];
        parameter.required = (/** @type {?} */ (obj['required']));
        if (obj['parameterType']) {
            /** @type {?} */
            var kind = obj['parameterType']['kind'];
            if (!kind) {
                throw new Error('解析参数数据类型时TypeKind为空！');
            }
            parameter.parameterType = DataTypeConvertUtil.initFromObject(obj['parameterType'], ctxt);
        }
        if (obj['constraint']) {
            /** @type {?} */
            var kind = obj['constraint']['kind'];
            if (!kind) {
                throw new Error('解析参数约束类型时ConstraintKind为空！');
            }
            parameter.constraint = ctxt.getConstraintSerializerFactory().get(kind).deserialize(obj['constraint'], ctxt);
        }
        if (obj["extension"]) {
            /** @type {?} */
            var extensionConvertor = new ExtensionConvertor();
            extensionConvertor.addToObj(parameter, obj);
        }
    };
    return ParameterConvertor;
}());
export { ParameterConvertor };
