// @farris/external-integration/esm5/lib/external-integration.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/external-integration.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from "@angular/core";
import { ExtIntegrationDataMappingDirective } from "./directives/ext-integration-data-mapping.directive";
import { CommonModule } from "@angular/common";
import { DialogService, FarrisDialogModule } from "@farris/ui-dialog";
import { BsModalService, ModalModule } from "@farris/ui-modal";
import { HttpClientModule } from "@angular/common/http";
import { FormsModule, ReactiveFormsModule } from "@angular/forms";
import { LocaleModule } from "@farris/ui-locale";
import { PerfectScrollbarModule } from "@farris/ui-perfect-scrollbar";
import { TreeModule } from "@farris/ui-tree";
import { FarrisTabsModule } from "@farris/ui-tabs";
import { FarrisCommonModule } from "@farris/ui-common";
import { FarrisFormsModule } from "@farris/ui-forms";
import { ExternalIntegrationComponent } from "./component/external-integration/external-integration.component";
import { TextModule } from "@farris/ui-text";
import { MessagerModule } from "@farris/ui-messager";
import { NotifyModule } from "@farris/ui-notify";
import { LayoutModule } from "@farris/ui-layout";
import { DatagridComponent, DatagridModule } from '@farris/ui-datagrid';
import { DatagridExtIntgrtnComponent } from "./component/external-integration/grid-external-integration.component";
import { InputGroupModule } from "@farris/ui-input-group";
import { FormMessageModule } from '@farris/ui-forms/validation-message';
import { ListNavModule } from '@farris/ui-list-nav';
import { ListViewModule } from '@farris/ui-list-view';
import { RouterModule } from "@angular/router";
var ExternalIntegrationModule = /** @class */ (function () {
    function ExternalIntegrationModule() {
    }
    ExternalIntegrationModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        ExtIntegrationDataMappingDirective,
                        ExternalIntegrationComponent,
                        DatagridExtIntgrtnComponent
                    ],
                    imports: [
                        CommonModule,
                        ReactiveFormsModule,
                        ListViewModule,
                        RouterModule,
                        ListNavModule,
                        InputGroupModule,
                        TextModule,
                        HttpClientModule,
                        FormMessageModule,
                        PerfectScrollbarModule,
                        TreeModule,
                        FarrisTabsModule,
                        FarrisFormsModule,
                        FormsModule,
                        FarrisCommonModule.forRoot(),
                        FarrisDialogModule.forRoot(),
                        MessagerModule.forRoot(),
                        NotifyModule.forRoot(),
                        LayoutModule,
                        LocaleModule.forRoot(),
                        ModalModule.forRoot(),
                        DatagridModule
                    ],
                    providers: [BsModalService, DialogService],
                    entryComponents: [DatagridExtIntgrtnComponent, DatagridComponent],
                    exports: [
                        ExtIntegrationDataMappingDirective,
                        ExternalIntegrationComponent,
                        DatagridExtIntgrtnComponent
                    ]
                },] }
    ];
    return ExternalIntegrationModule;
}());
export { ExternalIntegrationModule };
