// @gsp-svc/file-viewer/esm5/lib/image-viewer/image-viewer.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ImageViewerComponent } from './image-viewer/image-viewer.component';
import { ButtonsModule } from '@progress/kendo-angular-buttons';
import { MessagerModule } from '@farris/ui-messager';
import { ModalModule } from '@farris/ui-modal';
import { LoadingModule } from '@farris/ui-loading';
var ImageViewerModule = /** @class */ (function () {
    function ImageViewerModule() {
    }
    ImageViewerModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [ImageViewerComponent],
                    imports: [
                        CommonModule,
                        ButtonsModule,
                        MessagerModule.forRoot(),
                        ModalModule.forRoot(),
                        LoadingModule.forRoot({
                            message: 'Loading...'
                        }),
                    ],
                    exports: [ImageViewerComponent],
                    entryComponents: [ImageViewerComponent],
                },] }
    ];
    return ImageViewerModule;
}());
export { ImageViewerModule };
