// @farris/ui-datagrid-filter/esm2015/lib/filter-editors/filter-row-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FilterOperator, FilterConditionValue } from './../operations/operators';
import { Component, ViewChild, ElementRef, ChangeDetectorRef, Injector } from '@angular/core';
import { FilterRowEditorComponent } from './filter-row-editor.component';
import { DatagridFilterRowService } from '../datagrid-filter-row.service';
// <input [checked]="chekced" id="{{ id }}" type="checkbox" #chk >
export class FilterCheckboxComponent extends FilterRowEditorComponent {
    /**
     * @param {?} dfrs
     * @param {?} cd
     * @param {?} inject
     */
    constructor(dfrs, cd, inject) {
        super();
        this.dfrs = dfrs;
        this.cd = cd;
        this.inject = inject;
        this.i = 0;
        this.checked = false;
        this.dfrs.removeField.subscribe((/**
         * @param {?} f
         * @return {?}
         */
        f => {
            if (f === this.column.field) {
                this.clearValue();
            }
        }));
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.chk.nativeElement.indeterminate = true;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    checkedChange($event) {
        this.dfrs.closeFilterPanel();
        if (this.i > 0) {
            this.checked = !this.checked;
        }
        this.i += 1;
        if (this.i > 2) {
            this.i = 0;
            this.chk.nativeElement.indeterminate = true;
        }
        else {
            this.chk.nativeElement.indeterminate = false;
        }
        if (this.i === 0) {
            this.dfrs.updateColumnConditions(this.column.field, FilterConditionValue.All);
            this.checked = false;
        }
        else {
            /** @type {?} */
            const condition = {
                operator1: FilterOperator.Equal,
                value1: this.checked
            };
            this.dfrs.updateColumnConditions(this.column.field, condition);
        }
        this.cd.detectChanges();
        $event.stopPropagation();
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.checked = false;
        this.chk.nativeElement.indeterminate = true;
        this.cd.detectChanges();
    }
    /**
     * @param {?} filterCondition
     * @return {?}
     */
    initFilter(filterCondition) {
        if (filterCondition && Object.keys(filterCondition).length) {
            this.checked = filterCondition.value1;
            this.chk.nativeElement.indeterminate = false;
            this.i = this.checked ? 2 : 1;
        }
        else {
            this.checked = false;
            this.chk.nativeElement.indeterminate = true;
            this.i = 0;
        }
        this.dfrs._updateColumnConditions(this.column.field, filterCondition);
        this.cd.detectChanges();
    }
}
FilterCheckboxComponent.decorators = [
    { type: Component, args: [{
                selector: 'filter-checkbox',
                template: `
    <div class="custom-control custom-checkbox f-checkradio-single">
        <input [name]="column.field" [checked]="checked" type="checkbox" #chk class="custom-control-input">
        <label class="custom-control-label" (click)="checkedChange($event)"></label>
    </div>
    `
            }] }
];
/** @nocollapse */
FilterCheckboxComponent.ctorParameters = () => [
    { type: DatagridFilterRowService },
    { type: ChangeDetectorRef },
    { type: Injector }
];
FilterCheckboxComponent.propDecorators = {
    chk: [{ type: ViewChild, args: ['chk',] }]
};
if (false) {
    /** @type {?} */
    FilterCheckboxComponent.prototype.i;
    /** @type {?} */
    FilterCheckboxComponent.prototype.checked;
    /** @type {?} */
    FilterCheckboxComponent.prototype.chk;
    /** @type {?} */
    FilterCheckboxComponent.prototype.dfrs;
    /**
     * @type {?}
     * @private
     */
    FilterCheckboxComponent.prototype.cd;
    /**
     * @type {?}
     * @private
     */
    FilterCheckboxComponent.prototype.inject;
}
