// @farris/component-querycondition/esm5/lib/querycondition-advanced/config/compare-type.config.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/querycondition-advanced/config/compare-type.config.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var CompareType = {
    Equal: 0,
    NotEqual: 1,
    Greater: 2,
    GreaterOrEqual: 3,
    Less: 4,
    LessOrEqual: 5,
    Like: 6,
    LikeStartWith: 7,
    LikeEndWith: 8,
};
CompareType[CompareType.Equal] = 'Equal';
CompareType[CompareType.NotEqual] = 'NotEqual';
CompareType[CompareType.Greater] = 'Greater';
CompareType[CompareType.GreaterOrEqual] = 'GreaterOrEqual';
CompareType[CompareType.Less] = 'Less';
CompareType[CompareType.LessOrEqual] = 'LessOrEqual';
CompareType[CompareType.Like] = 'Like';
CompareType[CompareType.LikeStartWith] = 'LikeStartWith';
CompareType[CompareType.LikeEndWith] = 'LikeEndWith';
/** @type {?} */
var CompareTypeName = [
    '等于',
    '不等于',
    '大于',
    '大于等于',
    '小于',
    '小于等于',
    '包含',
    '开始是',
    '结束是',
];
/** @type {?} */
var CompareTypeInControl = [
    // Text = 0,
    [0, 6, 7, 8],
    // SingleDate = 1,
    [0, 2, 3, 4, 5],
    // SmartHelp = 2,
    [0, 1],
    // DropDownList = 3,
    [0, 1],
    // DateRange = 4,
    [],
    // NumberRange = 5,
    [],
    // SingleNumber = 6,
    [0, 1, 2, 3, 4, 5],
    // SingleYear = 7,
    [0, 2, 3, 4, 5],
    // BoolCheck = 8,
    [0],
    // DateTimeRange = 9,
    [],
    // SingleMonth = 10,
    [0, 2, 3, 4, 5],
    // MonthRange = 11,
    [],
    // SingleDateTime = 12,
    [0, 2, 3, 4, 5],
    // ComboLookUp = 13,
    [0, 1],
    // Radio = 14,
    [0],
    // InputGroup = 15
    [0, 6, 7, 8],
];
export { CompareType, CompareTypeName, CompareTypeInControl };
