// @farris/mobile-ui/lib/tag/index.js
import { openBlock, createBlock, Transition, withCtx, createVNode, renderSlot } from 'vue';

var script = {
  name: 'fm-tag',
  props: {
    closeable: {
      type: Boolean,
      "default": false
    },
    type: {
      type: String,
      "default": 'default'
    },
    className: {
      type: String
    },
    color: {
      type: String
    },
    sign: {
      type: Boolean,
      "default": false
    }
  },
  data: function data() {
    return {
      typeClass: '',
      signClass: ''
    };
  },
  mounted: function mounted() {
    this.typeClass = 'fm-tag-' + this.type;
    this.signClass = this.sign ? 'fm-tag-sign' : '';
  },
  watch: {
    type: {
      handler: function handler(newVal) {
        this.typeClass = 'fm-tag-' + newVal;
      }
    },
    sign: {
      handler: function handler(newVal) {
        this.signClass = newVal ? 'fm-tag-sign' : '';
      }
    }
  },
  computed: {
    styleObject: function styleObject() {
      var style = {
        background: this.color
      };
      return style;
    }
  },
  methods: {
    hexToRgba: function hexToRgba(hex, opacity) {
      if (opacity === void 0) {
        opacity = 1;
      }

      var reg = /^#([0-9a-fA-f]{3}|[0-9a-fA-f]{6})$/;
      var sColor = hex.toLowerCase();

      if (sColor && reg.test(sColor)) {
        if (sColor.length === 4) {
          var sColorNew = '#';

          for (var i = 1; i < 4; i += 1) {
            sColorNew += sColor.slice(i, i + 1).concat(sColor.slice(i, i + 1));
          }

          sColor = sColorNew;
        } //处理六位的颜色值


        var sColorChange = [];

        for (var i = 1; i < 7; i += 2) {
          sColorChange.push(parseInt('0x' + sColor.slice(i, i + 2)));
        }

        return 'RGB(' + sColorChange.join(',') + ',' + opacity + ')';
      } else {
        return sColor;
      }
    }
  }
};

function render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createBlock(Transition, {
    name: $props.closeable ? 'fm-fade' : null
  }, {
    "default": withCtx(function () {
      return [createVNode("span", {
        "class": ["fm-tag", [$props.className, $data.typeClass, $data.signClass]],
        style: $options.styleObject
      }, [renderSlot(_ctx.$slots, "default")], 6
      /* CLASS, STYLE */
      )];
    }),
    _: 3
    /* FORWARDED */

  }, 8
  /* PROPS */
  , ["name"]);
}

script.render = render;
script.__file = "packages/tag/src/index.vue";

script.install = function (app) {
  app.component(script.name, script);
};

export default script;
export { script as Tag };
