// @farris/devkit/esm5/lib/data-filter/object_filter_condition_data_filter.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { DataFilter } from './data_filter';
var ObjectFilterConditionDataFilter = /** @class */ (function (_super) {
    tslib_1.__extends(ObjectFilterConditionDataFilter, _super);
    function ObjectFilterConditionDataFilter() {
        return _super.call(this) || this;
    }
    ObjectFilterConditionDataFilter.prototype.filter = function (data, conditions) {
        var _this = this;
        if (data && data.size) {
            return data.filter(function (row) { return _this.validateRowData(row, conditions); });
        }
        return data;
    };
    ObjectFilterConditionDataFilter.prototype.validateRowData = function (bindingObject, conditions) {
        if (!conditions || Object.keys(conditions).length === 0) {
            return true;
        }
        return this.checkAllFieldInRowData(bindingObject, conditions);
    };
    ObjectFilterConditionDataFilter.prototype.checkAllFieldInRowData = function (bindingObject, conditions) {
        var _this = this;
        var resultObj = null;
        var currentLanguage = this.getCurrentLanguage();
        if (conditions) {
            var fields = Object.keys(conditions);
            resultObj = {};
            fields.forEach(function (field) {
                var fieldValue = _this.getValue(bindingObject, field);
                var property = bindingObject.properties.find(function (item) { return item.name === field; });
                if (property && property.enableMultiLangInput && fieldValue) {
                    fieldValue = fieldValue[currentLanguage];
                }
                var where = conditions[field];
                var value1 = where.value1, operator1 = where.operator1, relation = where.relation, operator2 = where.operator2, value2 = where.value2;
                if (where) {
                    var result = _this.getCompareResult(fieldValue, operator1, value1);
                    resultObj[field] = result;
                    if (relation) {
                        var result2 = _this.getCompareResult(fieldValue, operator2, value2);
                        resultObj[field] = _this.getRelationResult(result, relation, result2);
                    }
                }
            });
        }
        return this.checkAllFieldResult(resultObj);
    };
    ObjectFilterConditionDataFilter.prototype.getRelationResult = function (result1, relation, result2) {
        var _relation = relation.toLowerCase();
        if (_relation === 'and') {
            return result1 && result2;
        }
        else {
            return result1 || result2;
        }
    };
    ObjectFilterConditionDataFilter.prototype.checkAllFieldResult = function (resultObj) {
        if (!resultObj) {
            return true;
        }
        return Object.values(resultObj).reduce(function (c, n) {
            return c && n;
        }, true);
    };
    ObjectFilterConditionDataFilter.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ObjectFilterConditionDataFilter.ctorParameters = function () { return []; };
    return ObjectFilterConditionDataFilter;
}(DataFilter));
export { ObjectFilterConditionDataFilter };
