// @farris/devkit/esm5/lib/data-filter/types.js
export var Compare;
(function (Compare) {
    /**
     * 等于
     */
    Compare[Compare["Equal"] = 0] = "Equal";
    /**
     * 不等于
     */
    Compare[Compare["NotEqual"] = 1] = "NotEqual";
    /**
     * 大于
     */
    Compare[Compare["Greater"] = 2] = "Greater";
    /**
     * 大于等于
     */
    Compare[Compare["GreaterOrEqual"] = 3] = "GreaterOrEqual";
    /**
     * 小于
     */
    Compare[Compare["Less"] = 4] = "Less";
    /**
     * 小于等于
     */
    Compare[Compare["LessOrEqual"] = 5] = "LessOrEqual";
    /**
     * 包含
     */
    Compare[Compare["Like"] = 6] = "Like";
    /**
     * 不包含
     */
    Compare[Compare["NotLike"] = 9] = "NotLike";
    Compare[Compare["In"] = 14] = "In";
    /**
     * 为空
     */
    Compare[Compare["Empty"] = 1001] = "Empty";
    /**
     * 不为空
     */
    Compare[Compare["NotEmpty"] = 1002] = "NotEmpty";
})(Compare || (Compare = {}));
export var FilterRelation;
(function (FilterRelation) {
    FilterRelation[FilterRelation["Empty"] = 0] = "Empty";
    /**
     * 并且
     */
    FilterRelation[FilterRelation["And"] = 1] = "And";
    /**
     * 或者
     */
    FilterRelation[FilterRelation["Or"] = 2] = "Or";
})(FilterRelation || (FilterRelation = {}));
/**
 * 值类型
 */
export var ExpressValueType;
(function (ExpressValueType) {
    /**
     * 值
     */
    ExpressValueType[ExpressValueType["Value"] = 0] = "Value";
    /**
     * 表达式
     */
    ExpressValueType[ExpressValueType["Expression"] = 1] = "Expression";
    /**
     * 前端表达式
     */
    ExpressValueType["FrontExpress"] = "frontExpress";
})(ExpressValueType || (ExpressValueType = {}));
export var LeftBracket = [
    '', '(', '((', '((('
];
export var RighttBracket = [
    '', ')', '))', ')))'
];
