// @progress/kendo-angular-dialog/dist/es2015/window/actions/window-close-action.directive.js
import { Directive, HostListener, ElementRef, Renderer2, Optional, Input } from "@angular/core";
import { Button } from '@progress/kendo-angular-buttons';
import { DragResizeService } from '../drag-resize.service';
import { L10N_PREFIX, LocalizationService } from '@progress/kendo-angular-l10n';
export class WindowCloseActionDirective extends Button {
    constructor(el, renderer, _service, localization) {
        super(el, renderer, null, localization);
        this.window = _service;
        this.look = 'bare';
        this.icon = 'close';
    }
    onClick() {
        if (!this.isDisabled) {
            this.window.closeAction();
        }
    }
}
WindowCloseActionDirective.decorators = [
    { type: Directive, args: [{
                exportAs: 'kendoWindowCloseAction',
                providers: [
                    LocalizationService,
                    {
                        provide: L10N_PREFIX,
                        useValue: 'kendo.button'
                    }
                ],
                selector: 'button[kendoWindowCloseAction]' // tslint:disable-line
            },] },
];
/** @nocollapse */
WindowCloseActionDirective.ctorParameters = () => [
    { type: ElementRef, },
    { type: Renderer2, },
    { type: DragResizeService, decorators: [{ type: Optional },] },
    { type: LocalizationService, },
];
WindowCloseActionDirective.propDecorators = {
    'window': [{ type: Input },],
    'onClick': [{ type: HostListener, args: ['click',] },],
};
