// @progress/kendo-angular-dialog/dist/es2015/window/actions/window-maximize-action.directive.js
import { Directive, HostListener, ElementRef, Renderer2, Optional, HostBinding, Input } from "@angular/core";
import { DragResizeService } from './../drag-resize.service';
import { Button } from '@progress/kendo-angular-buttons';
import { L10N_PREFIX, LocalizationService } from '@progress/kendo-angular-l10n';
export class WindowMaximizeActionDirective extends Button {
    constructor(el, renderer, _service, localization) {
        super(el, renderer, null, localization);
        this.window = _service;
        this.look = 'bare';
        this.icon = 'window-maximize';
    }
    onClick() {
        if (!this.isDisabled) {
            this.window.maximizeAction();
        }
    }
    get visible() {
        return this.window.options.state === 'default' ? 'inline-flex' : 'none';
    }
}
WindowMaximizeActionDirective.decorators = [
    { type: Directive, args: [{
                exportAs: 'kendoWindowMaximizeAction',
                providers: [
                    LocalizationService,
                    {
                        provide: L10N_PREFIX,
                        useValue: 'kendo.button'
                    }
                ],
                selector: 'button[kendoWindowMaximizeAction]' // tslint:disable-line
            },] },
];
/** @nocollapse */
WindowMaximizeActionDirective.ctorParameters = () => [
    { type: ElementRef, },
    { type: Renderer2, },
    { type: DragResizeService, decorators: [{ type: Optional },] },
    { type: LocalizationService, },
];
WindowMaximizeActionDirective.propDecorators = {
    'window': [{ type: Input },],
    'onClick': [{ type: HostListener, args: ['click',] },],
    'visible': [{ type: HostBinding, args: ['style.display',] },],
};
