// @gsp-cmp/ubcc-param-data-ui/esm5/lib/ubcc-param-data-ui-dialog/ubcc-param-data-ui-dialog.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ubcc-param-data-ui-dialog/ubcc-param-data-ui-dialog.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input, TemplateRef, ViewChild } from '@angular/core';
import { UbccParamTemplate, UbccParamTemplateSerializer } from '@gsp-cmp/ubcc-param-vo';
import { UbccParamDataManangerService } from '../service/http-service';
import { DataTypeConst } from '@ecp-caf/common-structure';
import { UbccParamDataUtils } from '../utils/param-data-utils';
var UbccParamDataUiDialogComponent = /** @class */ (function () {
    function UbccParamDataUiDialogComponent(utils, http) {
        var _this = this;
        this.utils = utils;
        this.http = http;
        this.paramDataItems = [];
        this.confirmEmitter = new EventEmitter();
        this.closeEmitter = new EventEmitter();
        this.close = (/**
         * @return {?}
         */
        function () {
            if (_this.dialog) {
                _this.dialog.close();
            }
            _this.closeEmitter.emit();
        });
    }
    Object.defineProperty(UbccParamDataUiDialogComponent.prototype, "data", {
        /**
         * 输入数据为JSON对象
         */
        set: /**
         * 输入数据为JSON对象
         * @param {?} data
         * @return {?}
         */
        function (data) {
            this._data = data;
            this.loadParamData(this.templateId);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    UbccParamDataUiDialogComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.template = this.template || new UbccParamTemplate();
    };
    /**
     * @private
     * @param {?} templateId
     * @return {?}
     */
    UbccParamDataUiDialogComponent.prototype.loadParamData = /**
     * @private
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        var _this = this;
        if (!templateId) {
            return;
        }
        this.http.getParamTemplate(templateId).subscribe((/**
         * @param {?} vo
         * @return {?}
         */
        function (vo) {
            if (!vo.content) {
                return;
            }
            /** @type {?} */
            var serializer = new UbccParamTemplateSerializer();
            _this.template = serializer.deserializer(JSON.parse(vo.content));
            _this.getParamDataItemsStructure();
            _this.convertJsonToParamData();
        }));
    };
    /**
     * @private
     * @return {?}
     */
    UbccParamDataUiDialogComponent.prototype.convertJsonToParamData = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this._data || Object.keys(this._data).length <= 0) {
            return;
        }
        this.paramDataItems.map((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            /** @type {?} */
            var key = Object.keys(_this._data).find((/**
             * @param {?} x
             * @return {?}
             */
            function (x) { return x === item.code; }));
            if (!key || !_this._data[key]) {
                return;
            }
            if (_this.isCollection(item)) {
                //参数的值如果是数组，以逗号隔开
                item.value = _this._data[key].join(',');
                if (!_this.utils.isHelpType(item.parameterType)) {
                    return;
                }
                /** @type {?} */
                var helpId = ((/** @type {?} */ (_this.utils.getCollectionElementType(item.parameterType)))).refId;
                _this.http.getHelpData(helpId, _this._data[key]).subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                function (result) {
                    if (!result || result.length <= 0) {
                        return;
                    }
                    /** @type {?} */
                    var values = [];
                    result.map((/**
                     * @param {?} resItem
                     * @return {?}
                     */
                    function (resItem) { return values.push(resItem.text); }));
                    item.valueText = values.join(',');
                }));
            }
            else {
                item.value = _this._data[key];
                if (item.parameterTypeDesc !== 'help') {
                    return;
                }
                /** @type {?} */
                var helpId = ((/** @type {?} */ (item.parameterType))).refId;
                _this.http.getHelpData(helpId, [item.value]).subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                function (result) {
                    if (!result || result.length <= 0) {
                        return;
                    }
                    item.valueText = result[0].text;
                }));
            }
        }));
    };
    /**
     * @private
     * @return {?}
     */
    UbccParamDataUiDialogComponent.prototype.getParamDataItemsStructure = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.template || !this.template.params) {
            return;
        }
        this.paramDataItems = [];
        this.template.params.map((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return _this.paramDataItems.push(x); }));
    };
    /**
     * @private
     * @return {?}
     */
    UbccParamDataUiDialogComponent.prototype.covertParamDataToJson = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.paramDataItems || this.paramDataItems.length <= 0) {
            return null;
        }
        /** @type {?} */
        var object = new Object();
        this.paramDataItems.map((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (_this.isCollection(item)) {
                object[item.code] = [];
                /** @type {?} */
                var itemValues = item.value.split(',');
                if (!itemValues || itemValues.length <= 0) {
                    return;
                }
                itemValues.map((/**
                 * @param {?} itemValue
                 * @return {?}
                 */
                function (itemValue) { return object[item.code].push(itemValue); }));
                //帮助name赋值
                if (item.parameterTypeDesc === 'help') {
                    object[item.code + "_name"] = [];
                    /** @type {?} */
                    var itemTextValues = item.valueText.split(',');
                    if (!itemTextValues || itemTextValues.length <= 0) {
                        return;
                    }
                    itemTextValues.map((/**
                     * @param {?} itemTextValue
                     * @return {?}
                     */
                    function (itemTextValue) { return object[item.code + "_name"].push(itemTextValue); }));
                }
            }
            else {
                object[item.code] = item.value;
                //帮助name赋值
                if (item.parameterTypeDesc === 'help') {
                    object[item.code + "_name"] = item.valueText;
                }
            }
        }));
        return object;
    };
    /**
     * @return {?}
     */
    UbccParamDataUiDialogComponent.prototype.confirm = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var jsonData = this.covertParamDataToJson();
        this.confirmEmitter.emit(jsonData);
    };
    /**
     * @private
     * @param {?} param
     * @return {?}
     */
    UbccParamDataUiDialogComponent.prototype.isCollection = /**
     * @private
     * @param {?} param
     * @return {?}
     */
    function (param) {
        if (!param || !param.parameterType) {
            return false;
        }
        if (param.parameterType.getKind() === DataTypeConst.DEFAULT_COLLECTION) {
            return true;
        }
        else {
            return false;
        }
    };
    UbccParamDataUiDialogComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-ubcc-param-data-ui-dialog',
                    template: "<div class=\"d-flex\" style=\"position: absolute;top: 0;bottom: 0;left: 0;right: 0;padding: 0px 8px 0px 8px;\">\n  <lib-ubcc-param-data-ui class=\"f-utils-fill\" [(template)]=\"template\" [(paramDataItems)]=\"paramDataItems\"  style=\"overflow-y: auto;\"></lib-ubcc-param-data-ui>\n</div>\n\n<ng-template #btnRef>\n  <div style=\"width: 100%;line-height: 50px; text-align: center;min-width: 140px;\">\n    <button class=\"btn btn-primary\" (click)=confirm() style=\"float:right;margin-right: 5px\">\u786E\u5B9A</button>\n    <button class=\"btn btn-secondary\" (click)=close() style=\"float:right;margin-right: 5px\">\u5173\u95ED</button>\n  </div>\n</ng-template>",
                    styles: [":host{display:flex;flex-direction:column;flex:1}"]
                }] }
    ];
    /** @nocollapse */
    UbccParamDataUiDialogComponent.ctorParameters = function () { return [
        { type: UbccParamDataUtils },
        { type: UbccParamDataManangerService }
    ]; };
    UbccParamDataUiDialogComponent.propDecorators = {
        templateId: [{ type: Input }],
        data: [{ type: Input }],
        confirmEmitter: [{ type: Output }],
        closeEmitter: [{ type: Output }],
        btnRef: [{ type: ViewChild, args: ['btnRef',] }]
    };
    return UbccParamDataUiDialogComponent;
}());
export { UbccParamDataUiDialogComponent };
if (false) {
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.template;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype._data;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.templateId;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.paramDataItems;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.confirmEmitter;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.closeEmitter;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.btnRef;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.dialog;
    /** @type {?} */
    UbccParamDataUiDialogComponent.prototype.close;
    /**
     * @type {?}
     * @private
     */
    UbccParamDataUiDialogComponent.prototype.utils;
    /**
     * @type {?}
     * @private
     */
    UbccParamDataUiDialogComponent.prototype.http;
}
