// @farris/ui-tabs/esm2015/lib/tab/tab.transclude.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ViewContainerRef, TemplateRef, KeyValueDiffers } from '@angular/core';
export class TabTranscludeDirective {
    /**
     * @param {?} viewRef
     * @param {?} _differs
     */
    constructor(viewRef, _differs) {
        this._differs = _differs;
        // setTimeout(() => {
        this.viewRef = viewRef;
        // }, 5000);
        // this.cd.detectChanges();
    }
    /**
     * @return {?}
     */
    get trancludeTempl() {
        return this._trancludeTempl;
    }
    /**
     * @param {?} templateRef
     * @return {?}
     */
    set trancludeTempl(templateRef) {
        this._trancludeTempl = templateRef;
        if (templateRef) {
            this.viewRef.createEmbeddedView(templateRef);
        }
        // if (!this._differ && templateRef) {
        //     this._differ = this._differs.find(templateRef).create();
        // }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // this.setIntervalTemp = setInterval(() => {
        //     if (this._trancludeTempl && this.viewRef) {
        //         console.log('sa');
        //         // this.viewRef = this.viewRef;
        //         clearInterval(this.setIntervalTemp);
        //     }
        // }, 0);
    }
    /**
     * @return {?}
     */
    ngDoCheck() {
        // const CHANGE = this._differ.diff(this._trancludeTempl);
        // if (CHANGE) {
        //     this.viewRef.createEmbeddedView(this._trancludeTempl);
        // }
    }
}
TabTranscludeDirective.decorators = [
    { type: Directive, args: [{
                selector: '[tabTransclude]',
            },] }
];
/** @nocollapse */
TabTranscludeDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: KeyValueDiffers }
];
TabTranscludeDirective.propDecorators = {
    trancludeTempl: [{ type: Input, args: ['tabTransclude',] }]
};
if (false) {
    /** @type {?} */
    TabTranscludeDirective.prototype._trancludeTempl;
    /** @type {?} */
    TabTranscludeDirective.prototype.setIntervalTemp;
    /**
     * @type {?}
     * @private
     */
    TabTranscludeDirective.prototype._differ;
    /** @type {?} */
    TabTranscludeDirective.prototype.viewRef;
    /**
     * @type {?}
     * @private
     */
    TabTranscludeDirective.prototype._differs;
}
