// @farris/ide-publish-menu/esm2015/lib/publish-menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { cloneDeep } from 'lodash-es';
import { LoadingService } from '@farris/ui-loading';
import { MessagerService } from '@farris/ui-messager';
import { Component, Injector, Input, LOCALE_ID, ViewChild, TemplateRef, ChangeDetectorRef, ViewEncapsulation } from '@angular/core';
import { HttpService } from '@farris/ide-devkit';
import { BsModalService } from '@farris/ui-modal';
import { GspAppManagerComponent } from './gsp-app/gsp-app.component';
import { IdService } from '@farris/ui-common';
import { BizobjectOperatorsComponent } from './biz-object-operators/bizobject-operators.component';
import { SelectAppComponent } from './select-app/select-app.component';
import { of } from 'rxjs';
import { LocaleService } from '@farris/ui-locale';
import { CommonUtils } from '@farris/ui-common';
import { PUBLISH_MENU_LOCALE } from './locale';
export class PublishMenuComponent {
    /**
     * @param {?} http
     * @param {?} injector
     * @param {?} modalService
     * @param {?} uuid
     * @param {?} messagerSer
     * @param {?} loadingSer
     * @param {?} cd
     */
    constructor(http, injector, modalService, uuid, messagerSer, loadingSer, cd) {
        this.http = http;
        this.injector = injector;
        this.modalService = modalService;
        this.uuid = uuid;
        this.messagerSer = messagerSer;
        this.loadingSer = loadingSer;
        this.cd = cd;
        this.serverIp = '';
        this.publishPath = '';
        /**
         * 所有功能菜单数据
         */
        this.allFuncTreeData = [];
        this.funcIdField = 'id';
        this.showMask = true;
        this.languages = [
            { code: 'zh-CHS', name: '中文简体' },
            { code: 'en', name: 'English' },
            { code: 'zh-CHT', name: '中文繁体' }
        ];
        this.funcModel = null;
        this.localeId = 'zh-CHS';
        this.commonUtil = this.injector.get(CommonUtils);
        this.localeId = this.injector.get(LOCALE_ID, 'zh-CHS');
        if (this.localeId === 'zh') {
            this.localeId = 'zh-CHS';
        }
        if (!this.localeSer) {
            this.localeSer = this.injector.get(LocaleService, null);
            if (!this.localeSer) {
                this.localeSer = new LocaleService(this.localeId, this.commonUtil, PUBLISH_MENU_LOCALE);
            }
            else {
                if (!this.localeSer.langData['publishmenu']) {
                    this.localeSer.setLocaleData(PUBLISH_MENU_LOCALE);
                }
            }
        }
    }
    /**
     * @private
     * @return {?}
     */
    get FUNC_URI() {
        return this.serverIp + '/api/runtime/sys/v1.0/functions/allFuncs';
    }
    /**
     * @private
     * @return {?}
     */
    get APP_URI() {
        return this.serverIp + '/api/runtime/sys/v1.0/gspapp/all';
    }
    /**
     * @private
     * @return {?}
     */
    get APP_POST_URI() {
        return this.serverIp + '/api/runtime/sys/v1.0/gspapp';
    }
    /**
     * @private
     * @return {?}
     */
    get FUNC_POST_URI() {
        return this.serverIp + '/api/runtime/sys/v1.0/functions';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // this.initAllFuncTree();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
    }
    /**
     * @return {?}
     */
    emptyData() {
        this.funcModel = {
            appId: '',
            appInvokId: '',
            bizObjectId: '',
            bizOpId: '',
            code: '',
            creator: '',
            description: '',
            funcType: '4',
            id: this.uuid.generate(),
            icon: '',
            isDetail: true,
            isDisplayed: true,
            isSysInit: false,
            layer: '4',
            menuType: '1',
            name: '',
            parentId: '',
            path: '',
            staticParams: [{ name: 'mode', value: 'spa' }],
            url: '',
            nameLanguage: null
        };
    }
    /**
     * @return {?}
     */
    validate() {
        /** @type {?} */
        let msg = '';
        if (!this.funcModel.parentId) {
            msg += this.localeSer.getValue('publishmenu.func.message.parentId'); // '请在左侧选择上级菜单！<br>';
        }
        if (!msg) {
            if (!this.funcModel.code) {
                msg += this.localeSer.getValue('publishmenu.func.message.code'); // '功能编号不能为空！<br>';
            }
            if (!this.funcModel.name) {
                msg += this.localeSer.getValue('publishmenu.func.message.name'); // '功能名称不能为空！<br>';
            }
            if (!this.funcModel.bizOpId) {
                msg += this.localeSer.getValue('publishmenu.func.message.bizOpId'); // '请选择功能操作！ <br>';
            }
            if (!this.funcModel.appId) {
                msg += this.localeSer.getValue('publishmenu.func.message.appid'); // '请选择关联应用！<br>';
            }
            if (!this.funcModel.appInvokId) {
                msg += this.localeSer.getValue('publishmenu.func.message.appInvokId'); // '请选择关联应用入口！<br>';
            }
        }
        // if (!this.funcModel.bizObjectId) {
        //     msg += this.localeSer.getValue('publishmenu.func.message.bizObjectId');  // '请选择左侧业务操作对象！ <br>';
        // }
        if (msg) {
            this.messagerSer.show('warning', msg, { safeHtml: false, height: 300 });
            return false;
        }
        return true;
    }
    /**
     * @return {?}
     */
    saveMenu() {
        this.funcModel.name = this.funcModel.nameLanguage ? this.funcModel.nameLanguage[this.localeId] : '';
        if (this.validate()) {
            /** @type {?} */
            const staticParams = JSON.stringify(this.funcModel.staticParams);
            /** @type {?} */
            const _funcModel = cloneDeep(this.funcModel);
            _funcModel.staticParams = staticParams;
            _funcModel.name = _funcModel.nameLanguage[this.localeId];
            return this.http.post(this.FUNC_POST_URI, _funcModel);
        }
        return of(false);
    }
    /**
     * @return {?}
     */
    show() {
        /** @type {?} */
        const modalRef = this.modalService.show(this.dlgContent, {
            title: this.localeSer.getValue('publishmenu.dialog.title'),
            width: 760, height: 500,
            showMaxButton: false,
            resizable: false,
            enableScroll: false,
            buttons: [
                {
                    text: this.localeSer.getValue('publishmenu.dialog.btnOk'),
                    cls: 'btn btn-primary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        this.saveMenu().subscribe((/**
                         * @param {?} r
                         * @return {?}
                         */
                        r => {
                            if (r) {
                                this.messagerSer.success(this.localeSer.getValue('publishmenu.dialog.success'));
                                modalRef.close();
                                this.showMask = true;
                            }
                        }), (/**
                         * @param {?} err
                         * @return {?}
                         */
                        (err) => {
                            console.error(err);
                            if (err) {
                                /** @type {?} */
                                const error = err.error;
                                if (error) {
                                    /** @type {?} */
                                    const mesg = error.Message;
                                    this.messagerSer.warning(mesg);
                                }
                            }
                        }));
                    })
                },
                {
                    text: this.localeSer.getValue('publishmenu.dialog.btnCancel'),
                    cls: 'btn btn-secondary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        modalRef.close();
                        this.showMask = true;
                    })
                }
            ],
            opened: (/**
             * @return {?}
             */
            () => {
                this.initAllFuncTree();
            })
        });
        this.emptyData();
        return modalRef;
    }
    // 选择关联应用
    /**
     * @return {?}
     */
    showSelectAppModal() {
        /** @type {?} */
        const modalRef = this.modalService.show(SelectAppComponent, {
            title: this.localeSer.getValue('publishmenu.dialog.selectapp'),
            width: 760, height: 500,
            enableScroll: false,
            showMaxButton: false,
            resizable: false,
            initialState: {
                serverIp: this.serverIp,
                funcModel: this.funcModel
            },
            buttons: [
                {
                    text: this.localeSer.getValue('publishmenu.dialog.btnOk'),
                    cls: 'btn btn-primary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        /** @type {?} */
                        const sac = (/** @type {?} */ (modalRef.content));
                        /** @type {?} */
                        let msg = '';
                        if (sac.appId && sac.appInvokId) {
                            this.funcModel.appId = sac.appId;
                            this.funcModel.appInvokId = sac.appInvokId;
                            msg = '';
                        }
                        if (!sac.appId) {
                            msg = this.localeSer.getValue('publishmenu.dialog.selectInfo');
                        }
                        else if (!sac.appInvokId) {
                            msg = this.localeSer.getValue('publishmenu.dialog.appentry');
                        }
                        if (msg) {
                            this.messagerSer.warning(msg);
                        }
                        else {
                            modalRef.close();
                        }
                    })
                },
                {
                    text: this.localeSer.getValue('publishmenu.dialog.btnCancel'),
                    cls: 'btn btn-secondary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        modalRef.close();
                    })
                }
            ]
        });
        return modalRef;
    }
    // 选择业务操作
    /**
     * @return {?}
     */
    showBizObjectOperatorModal() {
        /** @type {?} */
        const modalRef = this.modalService.show(BizobjectOperatorsComponent, {
            title: this.localeSer.getValue('publishmenu.dialog.bizOptitle'),
            width: 760, height: 500,
            enableScroll: false,
            showMaxButton: false,
            resizable: false,
            initialState: {
                serverIp: this.serverIp,
                funcModel: this.funcModel
            },
            buttons: [
                {
                    text: this.localeSer.getValue('publishmenu.dialog.btnOk'),
                    cls: 'btn btn-primary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        /** @type {?} */
                        const boc = (/** @type {?} */ (modalRef.content));
                        /** @type {?} */
                        const bizObj = (/** @type {?} */ (boc.bizObjectId));
                        /** @type {?} */
                        const bizOp = (/** @type {?} */ (boc.bizOpId));
                        if (bizOp) {
                            /** @type {?} */
                            let bizObjID = '';
                            if (!bizObj) {
                                bizObjID = bizOp.data.businessId;
                            }
                            else {
                                bizObjID = bizObj.id;
                            }
                            if (bizObjID) {
                                this.funcModel.bizObjectId = bizObjID;
                                this.funcModel.bizOpId = bizOp.id;
                                modalRef.close();
                            }
                            else {
                                this.messagerSer.warning(this.localeSer.getValue('publishmenu.dialog.bizwarn'));
                            }
                        }
                        else {
                            this.messagerSer.warning(this.localeSer.getValue('publishmenu.dialog.bizaction'));
                        }
                    })
                },
                {
                    text: this.localeSer.getValue('publishmenu.dialog.btnCancel'),
                    cls: 'btn btn-secondary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        modalRef.close();
                    })
                }
            ]
        });
        return modalRef;
    }
    // 新建应用
    /**
     * @return {?}
     */
    showAppModal() {
        /** @type {?} */
        const modalRef = this.modalService.show(GspAppManagerComponent, {
            title: this.localeSer.getValue('publishmenu.dialog.app'),
            width: 760, height: 500,
            showMaxButton: false,
            resizable: false,
            enableScroll: false,
            initialState: {
                serverIp: this.serverIp,
                path: (this.publishPath || '').replace('index.htmlindex.html', 'index.html'),
                funcModel: this.funcModel
            },
            buttons: [
                {
                    text: this.localeSer.getValue('publishmenu.dialog.btnOk'),
                    cls: 'btn btn-primary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        /** @type {?} */
                        const amc = (/** @type {?} */ (modalRef.content));
                        if (amc.validate()) {
                            this.loadingSer.show();
                            this.http.post(this.APP_POST_URI, amc.appModel).subscribe((/**
                             * @param {?} r
                             * @return {?}
                             */
                            r => {
                                if (r) {
                                    this.funcModel.appId = amc.appModel.id;
                                    this.funcModel.appInvokId = amc.appModel.appInvoks[0].id;
                                    this.loadingSer.close();
                                    modalRef.close();
                                }
                            }));
                        }
                        // console.log(amc.appModel);
                    })
                },
                {
                    text: this.localeSer.getValue('publishmenu.dialog.btnCancel'),
                    cls: 'btn btn-secondary',
                    handle: (/**
                     * @return {?}
                     */
                    () => {
                        modalRef.close();
                    })
                }
            ]
        });
        return modalRef;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    treeNodeClicked(e) {
        this.funcModel.parentId = e.node.data.id;
        this.showMask = false;
        this.cd.detectChanges();
        /** @type {?} */
        const input = document.querySelector('#txtFuncCode');
        if (input) {
            input.click();
            input.focus();
        }
    }
    /**
     * @param {?} name
     * @return {?}
     */
    removeStaticParams(name) {
        this.funcModel.staticParams = this.funcModel.staticParams.filter((/**
         * @param {?} n
         * @return {?}
         */
        n => n.name !== name));
    }
    /**
     * @return {?}
     */
    addStaticParams() {
        this.funcModel.staticParams.push({ name: '', value: '' });
    }
    /**
     * @private
     * @return {?}
     */
    initAllFuncTree() {
        this.http.get(this.FUNC_URI).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            if (data) {
                this.allFuncTreeData = this.convert2TreeNodes(data);
            }
        }));
    }
    /**
     * @private
     * @param {?} data
     * @param {?=} parentId
     * @return {?}
     */
    convert2TreeNodes(data, parentId = '0') {
        if (data.length) {
            /** @type {?} */
            const nodes = data.filter((/**
             * @param {?} v
             * @return {?}
             */
            (v) => v.parentId === parentId));
            return nodes.map((/**
             * @param {?} item
             * @return {?}
             */
            item => {
                /** @type {?} */
                const node = {
                    data: item,
                    children: [],
                    selectable: item.layer === '3'
                };
                node.children = this.convert2TreeNodes(data, item.id);
                return node;
            }));
        }
        return [];
    }
}
PublishMenuComponent.decorators = [
    { type: Component, args: [{
                selector: 'publish-menu',
                template: "\r\n<ng-template #dlgContent>\r\n\r\n<div class=\"d-flex flex-row\" style=\"width: 760px; height: 400px;border:0px solid #ae00e6\">\r\n    <!-- Left -->\r\n    <div region=\"west\" class=\"west\" style=\"width: 280px;min-width:280px; height: 100%; background-color: #fff;border-right: 1px solid #e2e8f0;\">\r\n        <!-- \u83DC\u5355\u6811 -->\r\n        <div class=\"ps\" style=\"position: relative; height: 100%;\" [perfectScrollbar]=\"{}\">\r\n            <farris-treetable #tt=\"treeTable\" [data]=\"allFuncTreeData\" [idField]=\"funcIdField\" [striped]=\"false\" [fit]=\"true\"\r\n            [findField]=\"'name'\" (nodeSelected)=\"treeNodeClicked($event)\"\r\n            [singleSelect]=\"true\" [showIcon]=\"true\" [showFilterBar]=\"false\" [selectOnCheck]=\"true\" [checkOnSelect]=\"true\">\r\n                <ng-template farrisTemplate=\"body\" let-rowNode let-treeNode=\"node\" let-rowData=\"rowData\" >\r\n                    <tr #row=\"row\" [selectRow]=\"rowNode\" [selectRowDisabled]=\"!rowNode.node.selectable\" [dblclick]=\"true\">\r\n                        <td style=\"border: 0; cursor: pointer;\" [style.color]=\"row.selectRowDisabled? '#595959': '#000000'\">\r\n                            <farris-treeTableToggler [rowNode]=\"rowNode\"></farris-treeTableToggler>\r\n                            <span [innerHTML]=\"rowData['name'] | highlight: tt.findValue : 'name' : 'name'\"></span>\r\n                        </td>\r\n                    </tr>\r\n                </ng-template>\r\n            </farris-treetable>\r\n        </div>\r\n    </div>\r\n\r\n    <!-- main -->\r\n    <div region=\"center\" class=\"center flex-fill public-menu-main\" style=\"color: #000000;overflow: auto;\">\r\n        <div class=\"mengbanban\" *ngIf=\"showMask\" \r\n        style=\"position: absolute; z-index: 10; top: 0; width: 100%; height: 100%;opacity: 0.5; background: #fff; display: block; font-size: 40px; \">\r\n            <h1 style=\"margin-top: 145px; color: #000; margin-left: 20px; \">\u2190 {{ 'publishmenu.func.message.info' | locale }}</h1>\r\n        </div>\r\n        <div class=\"row\" style=\"margin: 10px 15px;\">\r\n            <div class=\"col-12\">\r\n                <div class=\"form-group\">\r\n                    <label for=\"txtFuncId\">{{ 'publishmenu.func.id' | locale: '\u529F\u80FD\u6807\u8BC6' }}</label>\r\n                    <input type=\"text\" class=\"form-control\" id=\"txtFuncId\" disabled [ngModel]=\"funcModel.id\">\r\n                </div>\r\n                <div class=\"form-group\">\r\n                    <label for=\"txtFuncCode\"><span class=\"farris-label-info text-danger\">*</span>{{ 'publishmenu.func.code' | locale: '\u529F\u80FD\u7F16\u53F7' }}</label>\r\n                    <input type=\"text\" class=\"form-control\" id=\"txtFuncCode\" [(ngModel)]=\"funcModel.code\">\r\n                </div>\r\n                <div class=\"form-group\">\r\n                    <label for=\"txtFuncName\"><span class=\"farris-label-info text-danger\">*</span>{{ 'publishmenu.func.name' | locale: '\u529F\u80FD\u540D\u79F0' }}</label>\r\n                    <!-- <input type=\"text\" class=\"form-control\" id=\"txtFuncName\" [(ngModel)]=\"funcModel.name\" > -->\r\n                    <language-textbox \r\n                    [languages]=\"languages\" [(ngModel)]=\"funcModel.nameLanguage\"\r\n                    [enableClear]=\"false\"\r\n                    [readonly]=\"false\"\r\n                    [disabled]=\"false\"\r\n                    [autoFocus]=\"true\"\r\n                    ></language-textbox>\r\n                </div>\r\n                <div class=\"form-group\">\r\n                    <label for=\"txtFuncAction\"><span class=\"farris-label-info text-danger\">*</span>{{ 'publishmenu.func.action' | locale: '\u529F\u80FD\u64CD\u4F5C' }}</label>\r\n\r\n                    <div class=\"input-group\">\r\n                        <input type=\"text\" class=\"form-control\" [(ngModel)]=\"funcModel.bizOpId\" readonly >\r\n                        <div class=\"input-group-append\">\r\n                            <button class=\"btn btn-outline-secondary\" (click)=\"showBizObjectOperatorModal()\" type=\"button\">{{ 'publishmenu.func.selectBiz' | locale: '\u9009\u62E9' }}</button>\r\n                        </div>\r\n                    </div>\r\n\r\n                </div>\r\n                <div class=\"form-group\">\r\n                    <label for=\"txtFuncRemark\">{{ 'publishmenu.func.remark' | locale: '\u529F\u80FD\u63CF\u8FF0' }}</label>\r\n                    <input type=\"text\" class=\"form-control\" id=\"txtFuncRemark\" [(ngModel)]=\"funcModel.description\" >\r\n                </div>\r\n\r\n                <div class=\"form-group\">\r\n                    <label><span class=\"farris-label-info text-danger\">*</span>{{ 'publishmenu.func.linkapp' | locale: '\u5173\u8054\u5E94\u7528' }}</label>\r\n                    <div class=\"input-group\">\r\n                        <input type=\"text\" class=\"form-control\" [(ngModel)]=\"funcModel.appId\" readonly>\r\n                        <div class=\"input-group-append\">\r\n                            <button class=\"btn btn-outline-secondary\" type=\"button\" (click)=\"showSelectAppModal()\">{{ 'publishmenu.func.selectapp' | locale: '\u9009\u62E9' }}</button>\r\n                            <button class=\"btn btn-outline-secondary\" type=\"button\" (click)=\"showAppModal()\">{{ 'publishmenu.func.addapp' | locale: '\u65B0\u5EFA' }}</button>\r\n                        </div>\r\n                    </div>\r\n                </div>\r\n                <div class=\"form-group\">\r\n                    <label for=\"txtappInvokid\"><span class=\"farris-label-info text-danger\">*</span>{{ 'publishmenu.func.appentry' | locale: '\u5E94\u7528\u5165\u53E3' }}</label>\r\n                    <input type=\"text\" class=\"form-control\" id=\"txtappInvokid\" [ngModel]=\"funcModel.appInvokId\" disabled>\r\n                </div>\r\n\r\n                <div class=\"form-group\">\r\n                    <label for=\"txtAppRemark\">{{ 'publishmenu.func.appParam' | locale: '\u53C2\u6570\u8BBE\u7F6E' }}</label>\r\n                    <table class=\"table table-bordered\">\r\n                        <colgroup>\r\n                            <col width=\"140\">\r\n                            <col width=\"200\">\r\n                            <col width=\"\">\r\n                        </colgroup>\r\n                        <thead>\r\n                            <tr>\r\n                                <td>\r\n                                    {{ 'publishmenu.func.paramName' | locale: '\u53C2\u6570\u540D\u79F0' }}\r\n                                </td>\r\n                                <td>\r\n                                    {{ 'publishmenu.func.paramValue' | locale: '\u53C2\u6570\u503C' }}\r\n                                </td>\r\n                                <td align=\"center\">\r\n                                    {{ 'publishmenu.func.paramAction' | locale: '\u64CD\u4F5C' }}\r\n                                </td>\r\n                            </tr>\r\n                        </thead>\r\n                        <tbody>\r\n                            <tr *ngFor=\"let n of funcModel.staticParams\">\r\n                                <td>\r\n                                    <input type=\"text\" [(ngModel)]=\"n.name\" class=\"form-control\">\r\n                                </td>\r\n                                <td>\r\n                                    <input type=\"text\" class=\"form-control\" [(ngModel)]=\"n.value\" >\r\n                                </td>\r\n                                <td style=\"padding-left: 0; padding-right: 0; text-align: center\">\r\n                                    <button type=\"button\" (click)=\"removeStaticParams(n.name)\" class=\"btn\">\r\n                                        <span class=\"k-icon k-i-minus\"></span></button>\r\n                                </td>\r\n                            </tr>\r\n                        </tbody>\r\n                        <tfoot>\r\n                            <tr>\r\n                                <td colspan=\"3\">\r\n                                    <button type=\"button\" (click)=\"addStaticParams()\" class=\"btn btn-primary\">\r\n                                        <span class=\"k-icon k-i-plus\"></span> {{ 'publishmenu.func.addParam' | locale: '\u65B0\u589E\u53C2\u6570' }}</button>\r\n                                </td>\r\n                            </tr>\r\n                        </tfoot>\r\n                    </table>\r\n\r\n                </div>\r\n\r\n                <div class=\"form-group\">\r\n                    <label for=\"txtFuncRemark\">ParentId</label>\r\n                    <input type=\"text\" class=\"form-control\" id=\"txtFuncRemark\" [ngModel]=\"funcModel.parentId\" disabled>\r\n                </div>\r\n                    \r\n            </div>\r\n        </div>\r\n\r\n    </div>\r\n</div>\r\n\r\n\r\n</ng-template>",
                exportAs: 'publishMenu',
                encapsulation: ViewEncapsulation.None,
                styles: [`
        .public-menu-main::-webkit-scrollbar-track-piece{width:8px;background-color:#e2e2e2;}
        .public-menu-main::-webkit-scrollbar{width:8px;height:6px;}
        .public-menu-main::-webkit-scrollbar-thumb{height:50px;background:#9fa1a0; cursor:pointer;}
        .public-menu-main::-webkit-scrollbar-thumb:hover{background:#686868; cursor:pointer
        `]
            }] }
];
/** @nocollapse */
PublishMenuComponent.ctorParameters = () => [
    { type: HttpService },
    { type: Injector },
    { type: BsModalService },
    { type: IdService },
    { type: MessagerService },
    { type: LoadingService },
    { type: ChangeDetectorRef }
];
PublishMenuComponent.propDecorators = {
    serverIp: [{ type: Input }],
    dlgContent: [{ type: ViewChild, args: ['dlgContent',] }],
    publishPath: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PublishMenuComponent.prototype.serverIp;
    /** @type {?} */
    PublishMenuComponent.prototype.dlgContent;
    /** @type {?} */
    PublishMenuComponent.prototype.publishPath;
    /**
     * 所有功能菜单数据
     * @type {?}
     */
    PublishMenuComponent.prototype.allFuncTreeData;
    /** @type {?} */
    PublishMenuComponent.prototype.funcIdField;
    /** @type {?} */
    PublishMenuComponent.prototype.showMask;
    /** @type {?} */
    PublishMenuComponent.prototype.languages;
    /** @type {?} */
    PublishMenuComponent.prototype.funcModel;
    /** @type {?} */
    PublishMenuComponent.prototype.localeId;
    /** @type {?} */
    PublishMenuComponent.prototype.localeSer;
    /** @type {?} */
    PublishMenuComponent.prototype.commonUtil;
    /**
     * @type {?}
     * @private
     */
    PublishMenuComponent.prototype.http;
    /**
     * @type {?}
     * @private
     */
    PublishMenuComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    PublishMenuComponent.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    PublishMenuComponent.prototype.uuid;
    /**
     * @type {?}
     * @private
     */
    PublishMenuComponent.prototype.messagerSer;
    /**
     * @type {?}
     * @private
     */
    PublishMenuComponent.prototype.loadingSer;
    /**
     * @type {?}
     * @private
     */
    PublishMenuComponent.prototype.cd;
}
