// @gsp-cmp/ubcc-param-vo/esm2015/lib/vo/serializer/param-data-serializer/data-item-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/param-data-serializer/data-item-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DataItem } from '../../param-data-vo/data-item';
export class DataItemSerializer {
    /**
     * @param {?} dataItem
     * @return {?}
     */
    serialize(dataItem) {
        if (!dataItem) {
            return null;
        }
        /** @type {?} */
        let jobj = new Object();
        jobj['id'] = dataItem.id;
        jobj['dataCode'] = dataItem.dataCode;
        jobj['dataName'] = dataItem.dataName;
        jobj['dataDesc'] = dataItem.dataDesc;
        jobj['dataValue'] = dataItem.dataValue;
        jobj['dataType'] = dataItem.dataType;
        return jobj;
    }
    /**
     * @param {?} jobj
     * @return {?}
     */
    deserialize(jobj) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        let dataItem = new DataItem();
        dataItem.id = jobj['id'];
        dataItem.dataCode = jobj['dataCode'];
        dataItem.dataName = jobj['dataName'];
        dataItem.dataDesc = jobj['dataDesc'];
        dataItem.dataValue = jobj['dataValue'];
        dataItem.dataType = jobj['dataType'];
        return dataItem;
    }
}
