// @gsp-cmp/ubcc-param-vo/esm2015/lib/vo/serializer/param-data-serializer/dimension-data-row-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/param-data-serializer/dimension-data-row-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DimensionDataRow } from '../../param-data-vo/dimension-data-row';
import { DataItem } from '../../param-data-vo/data-item';
export class DimensionDataRowSerializer {
    /**
     * @param {?} dataRow
     * @return {?}
     */
    serialize(dataRow) {
        /** @type {?} */
        let jobj = new Object();
        if (!dataRow) {
            return null;
        }
        /** @type {?} */
        let dimensionDataRow = (/** @type {?} */ (dataRow));
        jobj['id'] = dimensionDataRow.id;
        jobj['kind'] = dimensionDataRow.kind;
        if (dimensionDataRow.dimensionDataItems) {
            jobj['dimensionDataItems'] = [];
            dimensionDataRow.dimensionDataItems.forEach((/**
             * @param {?} dimensionDataItem
             * @return {?}
             */
            dimensionDataItem => {
                jobj['dimensionDataItems'] = dimensionDataItem;
            }));
        }
        if (dimensionDataRow.paramDataItems) {
            jobj['paramDataItems'] = [];
            dimensionDataRow.paramDataItems.forEach((/**
             * @param {?} paramItem
             * @return {?}
             */
            paramItem => {
                jobj['paramDataItems'] = paramItem;
            }));
        }
        return jobj;
    }
    /**
     * @param {?} jobj
     * @return {?}
     */
    deserialize(jobj) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        let dataRow = new DimensionDataRow();
        dataRow.id = jobj['id'];
        if (jobj['dimensionDataItems']) {
            dataRow.dimensionDataItems = new Array();
            jobj['dimensionDataItems'].forEach((/**
             * @param {?} dimensionDataItemJobj
             * @return {?}
             */
            dimensionDataItemJobj => {
                dataRow.dimensionDataItems.push(Object.assign(new DataItem(), dimensionDataItemJobj));
            }));
        }
        if (jobj['paramDataItems']) {
            dataRow.paramDataItems = new Array();
            jobj['paramDataItems'].forEach((/**
             * @param {?} paramDataItemsJobj
             * @return {?}
             */
            paramDataItemsJobj => {
                dataRow.dimensionDataItems.push(Object.assign(new DataItem(), paramDataItemsJobj));
            }));
        }
        return dataRow;
    }
}
