// @farris/lookup-config/esm2015/lib/components/lookup-config-metadata.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/lookup-config-metadata.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input } from '@angular/core';
export class LookupConfigMetaDataComponent {
    constructor() {
        // metadataPath = '';
        this.HelpMetadata = 'HelpMetadata';
        this.isDeisgnerEnv = false;
        this.selectedMeta = new EventEmitter();
        this.closeModal = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} event
     * @return {?}
     */
    selectedMetadata(event) {
        this.selectedMeta.emit(event);
    }
    /**
     * @return {?}
     */
    closeModalEmitter() {
        this.closeModal.emit();
    }
}
LookupConfigMetaDataComponent.decorators = [
    { type: Component, args: [{
                selector: 'farris-lookup-config-metadata',
                template: "<div class=\"f-lookup-config-metadata\">\r\n    <div class=\"f-lookup-config-metadata-inner\">\r\n        <app-select-metadatart typeList=\"HelpMetadata\" (selectedMetadata)=\"selectedMetadata($event)\"\r\n            (closeModalEmitter)=\"closeModalEmitter()\">\r\n        </app-select-metadatart>\r\n    </div> \r\n</div>"
            }] }
];
/** @nocollapse */
LookupConfigMetaDataComponent.ctorParameters = () => [];
LookupConfigMetaDataComponent.propDecorators = {
    metadataPath: [{ type: Input }],
    isDeisgnerEnv: [{ type: Input }],
    selectedMeta: [{ type: Output, args: ['selectedMeta',] }],
    closeModal: [{ type: Output, args: ['closeModal',] }]
};
if (false) {
    /** @type {?} */
    LookupConfigMetaDataComponent.prototype.HelpMetadata;
    /** @type {?} */
    LookupConfigMetaDataComponent.prototype.metadataPath;
    /** @type {?} */
    LookupConfigMetaDataComponent.prototype.isDeisgnerEnv;
    /** @type {?} */
    LookupConfigMetaDataComponent.prototype.selectedMeta;
    /** @type {?} */
    LookupConfigMetaDataComponent.prototype.closeModal;
}
