// @farris/devkit/esm2015/lib/entity/metadata/entity_metadata_util.js
import { MetadataUtil } from '../../metadata/index';
import { NG_ENTITY } from './entity_decorator';
// import { NgProperty } from './ng_property';
// import { NG_FIELD,   NgFieldProperty   } from './ng_field';
// import { NG_OBJECT,  NgObjectProperty  } from './ng_object';
// import { NG_Dynamic, NgDynamicProperty } from './ng_dynamic';
// import { NG_LIST,    NgListProperty    } from './ng_list';
import { NG_FIELD, NG_OBJECT, NG_Dynamic, NG_LIST } from './field_decorator';
/**
 * 实体元数据工具类
 */
export class EntityMetadataUtil {
    /**
     * 获取所有属性
     * @todo：封装根据基类获取所有元数据的方法，解决重复代码
     */
    static getAllNgProperties(entityType) {
        const ngPlainProperties = this.getNgFieldProperties(entityType);
        const ngEntityProperties = this.getNgObjectProperties(entityType);
        const ngDynamicProperties = this.getNgDynamicProperties(entityType);
        const ngEntityListProperties = this.getNgObjectProperties(entityType);
        return Object.assign({}, ngPlainProperties, ngEntityProperties, ngDynamicProperties, ngEntityListProperties);
    }
    /**
     * 获取实体元数据描述
     */
    static getNgEntityMatadata(entityType) {
        return MetadataUtil.getClassMetadataByNameWithTranslate(entityType, NG_ENTITY);
    }
    /**
     * 获取NgField的属性元数据
     */
    static getNgFieldProperties(entityType) {
        return MetadataUtil.getPropsMetadatasByName(entityType, NG_FIELD);
    }
    /**
     * 获取NgObject属性元数据
     */
    static getNgObjectProperties(entityType) {
        return MetadataUtil.getPropsMetadatasByName(entityType, NG_OBJECT);
    }
    /**
     * 获取NgDynamic属性元数据
     */
    static getNgDynamicProperties(entityType) {
        return MetadataUtil.getPropsMetadatasByName(entityType, NG_Dynamic);
    }
    /**
     * 获取NgList属性元数据
     */
    static getNgListProperties(entityType) {
        return MetadataUtil.getPropsMetadatasByName(entityType, NG_LIST);
    }
    /**
     * 获取主键属性元数据
     */
    static getPrimaryKeyProperty(entityType) {
        let primaryKeyProperty;
        const ngPlainProperties = EntityMetadataUtil.getNgFieldProperties(entityType);
        Object.keys(ngPlainProperties).forEach((propName) => {
            const ngProperty = ngPlainProperties[propName];
            if (ngProperty.primary === true) {
                primaryKeyProperty = ngProperty;
            }
        });
        return primaryKeyProperty;
    }
}
