// @qdp/ide-cmp/esm5/lib/component/schema-editor/schema-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, Output, EventEmitter, ViewChild, Injector } from '@angular/core';
import { FarrisTabsComponent } from '@farris/ui-tabs';
import { LocalizeService } from '@qdp/localize';
import { TitleEditorComponent } from './title-editor/title-editor.component';
import { ColumnEditorComponent } from './column-editor/column-editor.component';
import { StyleEditorComponent } from './style-editor/style-editor.component';
import { OtherEditorComponent } from './other-editor/other-editor.component';
var SchemaEditorComponent = /** @class */ (function () {
    function SchemaEditorComponent(injector) {
        this.injector = injector;
        this.tabsSelectChange = new EventEmitter();
        this.localizeService = this.injector.get(LocalizeService);
    }
    /**
     * @return {?}
     */
    SchemaEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    SchemaEditorComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
    };
    /**
     * @param {?} data
     * @return {?}
     */
    SchemaEditorComponent.prototype.handleTabSelect = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        /** @type {?} */
        var option = this.getOption();
        if (data.nextId === 'column') {
            this.columnOption.handleInfoChanged(0);
        }
        this.tabsSelectChange.emit(option);
    };
    /**
     * @return {?}
     */
    SchemaEditorComponent.prototype.getOption = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var option = {};
        if (this.titleOption) {
            option['titleOption'] = this.titleOption.getTitleOption();
        }
        if (this.columnOption) {
            option['columnOption'] = this.columnOption.getColumnOption();
        }
        if (this.styleOption) {
            option['styleOption'] = this.styleOption.getStyleOption();
        }
        if (this.otherOption) {
            option['otherOption'] = this.otherOption.getOtherOption();
        }
        return option;
    };
    SchemaEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'schema-editor',
                    template: "<farris-tabs [value]=\"'title'\"  [contentFill]=\"true\" [autoTitleWidth]=\"true\"  (tabChange)=\"handleTabSelect($event)\">\r\n  <farris-tab [title]=\"localizeService.getValue('ideCmp.schemaManager.titleOption.tab')\" [id]=\"'title'\">\r\n    <ng-template #content>\r\n      <title-editor [currSchema]=\"currSchema\" class=\"schema-editor-tab d-flex flex-column flex-fill\"></title-editor>\r\n    </ng-template>\r\n  </farris-tab>\r\n  <farris-tab [title]=\"localizeService.getValue('ideCmp.schemaManager.columnOption.tab')\" [id]=\"'column'\">\r\n    <ng-template #content>\r\n      <column-editor [currSchema]=\"currSchema\" [colList]=\"colList\" [controlType]=\"controlType\"  class=\"schema-editor-tab d-flex flex-column flex-fill\"></column-editor>\r\n    </ng-template>\r\n  </farris-tab>\r\n  <farris-tab [title]=\"localizeService.getValue('ideCmp.schemaManager.styleOption.tab')\" [id]=\"'style'\">\r\n    <ng-template #content>\r\n      <style-editor [currSchema]=\"currSchema\" class=\"schema-editor-tab d-flex flex-column flex-fill\"></style-editor>\r\n    </ng-template>\r\n  </farris-tab>\r\n  <farris-tab [title]=\"localizeService.getValue('ideCmp.schemaManager.otherOption.tab')\" [id]=\"'other'\">\r\n    <ng-template #content>\r\n      <other-editor [currSchema]=\"currSchema\"  [controlType]=\"controlType\"  class=\"schema-editor-tab d-flex flex-column flex-fill\"></other-editor>\r\n    </ng-template>\r\n  </farris-tab>\r\n</farris-tabs>\r\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".schema-editor-tab{position:absolute;top:0;bottom:0;left:0;right:0;overflow:hidden}"]
                }] }
    ];
    /** @nocollapse */
    SchemaEditorComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    SchemaEditorComponent.propDecorators = {
        currSchema: [{ type: Input }],
        colList: [{ type: Input }],
        controlType: [{ type: Input }],
        tabsSelectChange: [{ type: Output }],
        titleOption: [{ type: ViewChild, args: [TitleEditorComponent,] }],
        columnOption: [{ type: ViewChild, args: [ColumnEditorComponent,] }],
        styleOption: [{ type: ViewChild, args: [StyleEditorComponent,] }],
        otherOption: [{ type: ViewChild, args: [OtherEditorComponent,] }],
        tabs: [{ type: ViewChild, args: [FarrisTabsComponent,] }]
    };
    return SchemaEditorComponent;
}());
export { SchemaEditorComponent };
if (false) {
    /** @type {?} */
    SchemaEditorComponent.prototype.currSchema;
    /** @type {?} */
    SchemaEditorComponent.prototype.colList;
    /** @type {?} */
    SchemaEditorComponent.prototype.controlType;
    /** @type {?} */
    SchemaEditorComponent.prototype.editorParams;
    /** @type {?} */
    SchemaEditorComponent.prototype.tabsSelectChange;
    /** @type {?} */
    SchemaEditorComponent.prototype.titleOption;
    /** @type {?} */
    SchemaEditorComponent.prototype.columnOption;
    /** @type {?} */
    SchemaEditorComponent.prototype.styleOption;
    /** @type {?} */
    SchemaEditorComponent.prototype.otherOption;
    /** @type {?} */
    SchemaEditorComponent.prototype.tabs;
    /** @type {?} */
    SchemaEditorComponent.prototype.localizeService;
    /**
     * @type {?}
     * @private
     */
    SchemaEditorComponent.prototype.injector;
}
