// @farris/ide-devkit/esm2015/lib/ideframework/eventbus/duplex-event-pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { forkJoin, Observable, of, Subject, timer } from 'rxjs';
import { mapTo } from 'rxjs/operators';
export class DuplexEventPipe {
    /**
     * @param {?} eventName
     */
    constructor(eventName) {
        this.listenerResults = new Map();
        this.subject = new Subject();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    notify(value) {
        /** @type {?} */
        let thenFn = () => { };
        /** @type {?} */
        let catchFn = () => { };
        /** @type {?} */
        const timestamp = Date.now().toString();
        this.subject.next({ value, timestamp });
        // fooObservable无意义，有两个作用：
        // 1. 没有实际observable时能够让forkJoin输出值，进入subscriber.next
        // 2. 建立延迟的observable（跳出当前micro task），避免subscription未返回和thenFn未被设置。
        /** @type {?} */
        const fooObservable = timer(0).pipe(mapTo(true));
        /** @type {?} */
        let listenerResultArray = this.listenerResults.get(timestamp);
        if (!listenerResultArray) {
            listenerResultArray = [fooObservable];
        }
        else {
            listenerResultArray = listenerResultArray.concat(fooObservable);
            this.listenerResults.delete(timestamp);
        }
        /** @type {?} */
        const subscription = forkJoin(listenerResultArray).subscribe({ next: (results) => {
                /** @type {?} */
                const success = results.reduce((previous, current) => previous && current);
                if (success) {
                    thenFn();
                }
                else {
                    catchFn();
                }
                subscription.unsubscribe();
            }, error: () => {
                subscription.unsubscribe();
            } });
        return {
            /**
             * @param {?} callback
             * @return {?}
             */
            then(callback) {
                thenFn = callback;
                return this;
            },
            /**
             * @param {?} callback
             * @return {?}
             */
            catch(callback) {
                catchFn = callback;
                return this;
            }
        };
    }
    /**
     * @param {?} callback
     * @param {?=} caller
     * @return {?}
     */
    listen(callback, caller) {
        /** @type {?} */
        const listener = caller ? callback.bind(caller) : callback;
        this.subject.subscribe(({ value, timestamp }) => {
            /** @type {?} */
            let listenerResult;
            /** @type {?} */
            const invokeResult = listener(value);
            if (invokeResult instanceof Observable) {
                listenerResult = invokeResult;
            }
            else if (typeof invokeResult === 'boolean') {
                listenerResult = of(invokeResult);
            }
            else {
                listenerResult = of(true);
            }
            /** @type {?} */
            let listenerResultArray = this.listenerResults.get(timestamp);
            if (!listenerResultArray) {
                listenerResultArray = [];
                this.listenerResults.set(timestamp, listenerResultArray);
            }
            listenerResultArray.push(listenerResult);
        });
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    DuplexEventPipe.prototype.listenerResults;
    /**
     * @type {?}
     * @private
     */
    DuplexEventPipe.prototype.subject;
}
