// @farris/ide-devkit/esm2015/lib/ideframework/eventbus/event-pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs/Subject';
export class EventPipe {
    /**
     * @param {?} name
     * @param {?} tokenValue
     * @param {?} emitter
     * @param {?} parentEventPipeList
     */
    constructor(name, tokenValue, emitter, parentEventPipeList) {
        this.name = name;
        this.tokenValue = tokenValue;
        this.emitter = emitter;
        this.parentEventPipeList = parentEventPipeList;
        this.eventSubject = new Subject();
        this.subscriptionMap = new Map();
        this.onceSubscriptionMap = new Map();
        if (this.parentEventPipeList) {
            this.parentEventPipeList.push(this);
        }
    }
    /**
     * @param {?} data
     * @return {?}
     */
    post(data) {
        this.eventSubject.next(data);
    }
    /**
     * @param {?} eventHandler
     * @param {?} caller
     * @return {?}
     */
    subscribe(eventHandler, caller) {
        /** @type {?} */
        const subscription = this.eventSubject.subscribe((value) => eventHandler.call(caller, value));
        this.subscriptionMap.set(caller, subscription);
        return this;
    }
    /**
     * @param {?} eventHandler
     * @param {?} caller
     * @return {?}
     */
    subscribeOnce(eventHandler, caller) {
        /** @type {?} */
        const subscription = this.eventSubject.subscribe((value) => eventHandler.call(caller, value));
        this.onceSubscriptionMap.set(caller, subscription);
        return this;
    }
    /**
     * @param {?} subscriber
     * @return {?}
     */
    unSubscribe(subscriber) {
        /** @type {?} */
        let subscription = this.subscriptionMap.get(subscriber);
        if (subscription) {
            subscription.unsubscribe();
            subscription = null;
            this.subscriptionMap.delete(subscriber);
        }
        else {
            subscription = this.onceSubscriptionMap.get(subscriber);
            if (subscription) {
                subscription.unsubscribe();
                subscription = null;
                this.onceSubscriptionMap.delete(subscriber);
            }
        }
    }
    // 注销使用once方法注册的订阅。
    /**
     * @return {?}
     */
    unSubscribeForOnce() {
        for (const subscriber of Array.from(this.onceSubscriptionMap.keys())) {
            this.unSubscribe(subscriber);
        }
    }
    /**
     * @param {?} emitter
     * @param {?} tokenValue
     * @return {?}
     */
    matchEmitterToken(emitter, tokenValue) {
        if (this.emitter && emitter && this.emitter !== emitter) {
            return false;
        }
        if (this.tokenValue && tokenValue && this.tokenValue !== tokenValue) {
            return false;
        }
        return true;
    }
    /**
     * @param {?} target
     * @param {?} tokenValue
     * @return {?}
     */
    examByTargetToken(target, tokenValue) {
        if (this.emitter !== target) {
            return false;
        }
        if (this.tokenValue !== tokenValue) {
            return false;
        }
        return true;
    }
    /**
     * @param {?} subscriber
     * @return {?}
     */
    dispose(subscriber) {
        this.unSubscribe(subscriber);
        if (this.subscriptionMap.size === 0 && this.parentEventPipeList) {
            /** @type {?} */
            const location = this.parentEventPipeList.findIndex(item => item === this);
            if (location !== -1) {
                this.parentEventPipeList.splice(location, 1);
            }
        }
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    EventPipe.prototype.eventSubject;
    /**
     * @type {?}
     * @private
     */
    EventPipe.prototype.subscriptionMap;
    /**
     * @type {?}
     * @private
     */
    EventPipe.prototype.onceSubscriptionMap;
    /**
     * @type {?}
     * @private
     */
    EventPipe.prototype.name;
    /**
     * @type {?}
     * @private
     */
    EventPipe.prototype.tokenValue;
    /**
     * @type {?}
     * @private
     */
    EventPipe.prototype.emitter;
    /**
     * @type {?}
     * @private
     */
    EventPipe.prototype.parentEventPipeList;
}
