// @farris/ui-filter-panel/esm5/lib/components/searchcomponent/search.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild } from '@angular/core';
import { InputGroupComponent } from '@farris/ui-input-group';
var PanelSearchComponent = /** @class */ (function () {
    function PanelSearchComponent() {
        this.searchHandle = new EventEmitter();
        this.submitChange = new EventEmitter();
        this.groupIcon = '<span class="k-icon k-i-search"></span>';
    }
    Object.defineProperty(PanelSearchComponent.prototype, "item", {
        get: /**
         * @return {?}
         */
        function () {
            return this._item;
        },
        set: /**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            this._item = data;
            if (this._item) {
                if (!this._item.hasOwnProperty('value')) {
                    this._item.value = { value: '' };
                }
                if (this._item.placeHolder) {
                    this.placeholder = this._item.placeHolder || '';
                }
                this.searchValue = this._item.value.value;
            }
            this.setSubmitChange();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PanelSearchComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.searchText) {
            this.searchText.textbox.nativeElement.focus();
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    PanelSearchComponent.prototype.onSearch = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this._item.value.value = this.searchValue;
        this._item.valueText = this._item.value.value;
        this.setSubmitChange();
        this.searchHandle.emit($event);
    };
    /**
     * @return {?}
     */
    PanelSearchComponent.prototype.setSubmitChange = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var val = this._item.value;
        if (Object.keys(val).length < 0 || !val.value) {
            this.submitChange.emit(false);
        }
        else {
            this.submitChange.emit(true);
        }
    };
    PanelSearchComponent.decorators = [
        { type: Component, args: [{
                    selector: 'panel-filter-search',
                    template: "\r\n<input-group [groupText]=\"groupIcon\" [(ngModel)]=\"searchValue\" (ngModelChange)=\"onSearch($event)\" (clear)=\"onSearch($event)\"\r\n(blurHandle)=\"onSearch($event)\" (enterHandle)=\"onSearch($event)\" (clickHandle)=\"onSearch($event)\"\r\n[placeholder]=\"placeholder\" #searchText></input-group>"
                }] }
    ];
    /** @nocollapse */
    PanelSearchComponent.ctorParameters = function () { return []; };
    PanelSearchComponent.propDecorators = {
        item: [{ type: Input }],
        searchHandle: [{ type: Output }],
        submitChange: [{ type: Output }],
        searchText: [{ type: ViewChild, args: ['searchText',] }]
    };
    return PanelSearchComponent;
}());
export { PanelSearchComponent };
if (false) {
    /** @type {?} */
    PanelSearchComponent.prototype._item;
    /** @type {?} */
    PanelSearchComponent.prototype.searchHandle;
    /** @type {?} */
    PanelSearchComponent.prototype.submitChange;
    /** @type {?} */
    PanelSearchComponent.prototype.searchText;
    /** @type {?} */
    PanelSearchComponent.prototype.placeholder;
    /** @type {?} */
    PanelSearchComponent.prototype.groupIcon;
    /** @type {?} */
    PanelSearchComponent.prototype.searchValue;
}
