// @farris/ui-response-toolbar/esm2015/lib/model/response-toolbar.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class ResponseToolbarBasic {
    // 记录宽度
    /**
     * @param {?} item
     */
    constructor(item) {
        /** @type {?} */
        const defaultData = {
            id: '',
            class: 'btn-secondary',
            icon: '',
            groupId: '',
            order: -1,
            asDropDownTop: false,
            text: '',
            isDP: false,
            disabled: false,
            visible: true,
            responsed: false,
            width: 0,
            tipsEnable: false,
            tipsText: ''
        };
        Object.keys(defaultData).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        propName => {
            this[propName] = item.hasOwnProperty(propName)
                ? item[propName]
                : defaultData[propName];
        }));
    }
    // 设置宽度
    /**
     * @param {?} value
     * @return {?}
     */
    setWidth(value) {
        this.width = parseInt(value, 10);
    }
    // 获取宽度
    /**
     * @return {?}
     */
    getWidth() {
        if (this.visible) {
            return this.width;
        }
        return false;
    }
}
if (false) {
    /** @type {?} */
    ResponseToolbarBasic.prototype.id;
    /** @type {?} */
    ResponseToolbarBasic.prototype.class;
    /** @type {?} */
    ResponseToolbarBasic.prototype.icon;
    /** @type {?} */
    ResponseToolbarBasic.prototype.groupId;
    /** @type {?} */
    ResponseToolbarBasic.prototype.order;
    /** @type {?} */
    ResponseToolbarBasic.prototype.asDropDownTop;
    /** @type {?} */
    ResponseToolbarBasic.prototype.text;
    /** @type {?} */
    ResponseToolbarBasic.prototype.isDP;
    /** @type {?} */
    ResponseToolbarBasic.prototype.disabled;
    /** @type {?} */
    ResponseToolbarBasic.prototype.visible;
    /** @type {?} */
    ResponseToolbarBasic.prototype.responsed;
    /** @type {?} */
    ResponseToolbarBasic.prototype.tipsEnable;
    /** @type {?} */
    ResponseToolbarBasic.prototype.tipsText;
    /** @type {?} */
    ResponseToolbarBasic.prototype.width;
}
export class ResponseToolbarItem extends ResponseToolbarBasic {
    // 点击事件
    /**
     * @param {?} item
     */
    constructor(item) {
        super(item);
        /** @type {?} */
        const defaultData = {
            text: '',
            event: null,
            isDP: false
        };
        Object.keys(defaultData).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        propName => {
            this[propName] = item.hasOwnProperty(propName)
                ? item[propName]
                : defaultData[propName];
        }));
    }
}
if (false) {
    /** @type {?} */
    ResponseToolbarItem.prototype.event;
}
export class ResponseToolbarDropDown extends ResponseToolbarBasic {
    /**
     * @param {?} item
     */
    constructor(item) {
        super(item);
        /** @type {?} */
        const defaultData = {
            topItem: null,
            isDP: true,
            children: [],
            class: 'btn-secondary',
            dropdownCls: '',
            menuCls: '',
            placement: '',
            split: false
        };
        Object.keys(defaultData).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        propName => {
            this[propName] = item.hasOwnProperty(propName)
                ? item[propName]
                : defaultData[propName];
        }));
    }
    /**
     * 是否有子项
     * @return {?}
     */
    hasChild() {
        return this.children.length > 0 ? true : false;
    }
    /**
     * 追加子项,默认从开头插入
     * @param {?} item
     * @return {?}
     */
    addChild(item) {
        this.children.unshift(item);
    }
    // 从头部开始删除子元素
    /**
     * @return {?}
     */
    removeChild() {
        this.children.shift();
    }
}
if (false) {
    /** @type {?} */
    ResponseToolbarDropDown.prototype.placement;
    /** @type {?} */
    ResponseToolbarDropDown.prototype.topItem;
    /** @type {?} */
    ResponseToolbarDropDown.prototype.dropdownCls;
    /** @type {?} */
    ResponseToolbarDropDown.prototype.menuCls;
    /** @type {?} */
    ResponseToolbarDropDown.prototype.split;
    /** @type {?} */
    ResponseToolbarDropDown.prototype.children;
}
export class ResponseToolbarGroup {
    // 记录宽度
    /**
     * @param {?} _id
     * @param {?} _groupName
     */
    constructor(_id, _groupName) {
        this.presetId = []; // 记录元素的Id
        // 记录元素的Id
        this.responsedIndex = []; // 记录转变为下拉时元素的位置
        this.id = _id;
        this.name = _groupName;
        this.width = 0;
    }
    // 设置宽度
    /**
     * @param {?} value
     * @return {?}
     */
    setWidth(value) {
        this.width = parseInt(value + '', 10);
    }
    // 获取宽度
    /**
     * @return {?}
     */
    getWidth() {
        return this.width;
    }
    // 更新presetIndex
    /**
     * @param {?} value
     * @return {?}
     */
    setPreset(value) {
        if (Array.isArray(value)) {
            this.presetId = this.presetId.concat(value);
        }
        else {
            this.presetId.push(value);
        }
    }
    // 清除preset
    /**
     * @return {?}
     */
    delPreset() {
        this.presetId = [];
    }
    // 删除
    /**
     * @param {?} index
     * @return {?}
     */
    removeResponsed(index) {
        this.responsedIndex.splice(index, 1);
    }
    // 更新responsedIndex
    /**
     * @param {?} value
     * @return {?}
     */
    setResponsed(value) {
        if (Array.isArray(value)) {
            this.responsedIndex = this.responsedIndex.concat(value);
        }
        else {
            this.responsedIndex.push(value);
        }
    }
    // 是否已经开始处理响应式
    /**
     * @return {?}
     */
    isResponsing() {
        return this.responsedIndex.length > 0;
    }
    // 是否已经处理完响应式
    /**
     * @return {?}
     */
    isResponsed() {
        return this.presetId.length === this.responsedIndex.length;
    }
}
if (false) {
    /** @type {?} */
    ResponseToolbarGroup.prototype.id;
    /** @type {?} */
    ResponseToolbarGroup.prototype.name;
    /** @type {?} */
    ResponseToolbarGroup.prototype.presetId;
    /** @type {?} */
    ResponseToolbarGroup.prototype.responsedIndex;
    /** @type {?} */
    ResponseToolbarGroup.prototype.width;
}
/**
 * @record
 */
export function ResponseToolbarClickEvent() { }
if (false) {
    /** @type {?} */
    ResponseToolbarClickEvent.prototype.id;
    /** @type {?} */
    ResponseToolbarClickEvent.prototype.text;
    /** @type {?} */
    ResponseToolbarClickEvent.prototype.hidden;
}
