// @gsp-wf/wf-process-editor/esm5/lib/views/backward-rule/backward-rule.entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BackWardRule = /** @class */ (function () {
    function BackWardRule() {
        this.ClrTypeID = BackWardRule.Clr_Type_ID;
    }
    BackWardRule.Clr_Type_ID = 'wf.BackWardRule';
    return BackWardRule;
}());
export { BackWardRule };
if (false) {
    /** @type {?} */
    BackWardRule.Clr_Type_ID;
    /** @type {?} */
    BackWardRule.prototype.ClrTypeID;
    /** @type {?} */
    BackWardRule.prototype.TargetRef;
    /** @type {?} */
    BackWardRule.prototype.SelectTargetAble;
    /** @type {?} */
    BackWardRule.prototype.WaitReturn;
    /** @type {?} */
    BackWardRule.prototype.DynamicBackWardRuleAble;
    /** @type {?} */
    BackWardRule.prototype.SelectTargetList;
    /** @type {?} */
    BackWardRule.prototype.ResubmitEffect;
}
/** @enum {string} */
var ResubmitEffect = {
    sequence: 'sequence',
    backToMe: 'backToMe',
    backToCurrentNode: 'backToCurrentNode',
};
export { ResubmitEffect };
