// @qdp/echarts/esm2015/lib/directive/bindingdata/form.bindingdata.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Optional, Injector } from '@angular/core';
import { BindingData, ViewModel } from '@farris/devkit';
import { NgxEchartsDirective } from '../qdp-charts/qdp.charts.directive';
import { QDPCacheUtil, RtfServices, SchemaManagerService } from '@qdp/common';
import { CacheService } from '@ecp-caf/caf-common';
export class FormBindingDataDirective {
    /**
     * @param {?} charts
     * @param {?} bindingData
     * @param {?} viewModel
     * @param {?} schemaManager
     * @param {?} injector
     */
    constructor(charts, bindingData, viewModel, schemaManager, injector) {
        this.charts = charts;
        this.bindingData = bindingData;
        this.viewModel = viewModel;
        this.schemaManager = schemaManager;
        this.injector = injector;
        this.delay = 0;
        this.interval = 0;
        if (this.injector) {
            this.cache = this.injector.get(CacheService);
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        QDPCacheUtil.setCache(RtfServices.getTabId(this.queryId) + this.queryId + 'form-charts', this);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.delayLoad();
    }
    /**
     * @return {?}
     */
    delayLoad() {
        if (!this.delay) {
            this.delay = 0;
        }
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.bindingListChange();
            this.intervalLoad();
        }), this.delay);
    }
    /**
     * @return {?}
     */
    intervalLoad() {
        if (this.interval) {
            this.myVar = setInterval((/**
             * @return {?}
             */
            () => {
                this.bindingListChange();
            }), this.interval);
        }
    }
    /**
     * @return {?}
     */
    clearIntervalLoad() {
        if (this.myVar) {
            clearInterval(this.myVar);
        }
    }
    /**
     * @private
     * @return {?}
     */
    bindingListChange() {
        if (this.bindingData) {
            this.bindData();
            this.bindingData.changes.subscribe((/**
             * @param {?} change
             * @return {?}
             */
            (change) => {
                this.bindData();
            }));
        }
    }
    /**
     * @private
     * @return {?}
     */
    bindData() {
        /** @type {?} */
        const self = this;
        if (this.bindingList) {
            /** @type {?} */
            let op;
            /** @type {?} */
            const organizationId = this.cache.get(RtfServices.getTabId(self.queryId) + 'organizationId') || '';
            this.schemaManager.getDefaultSchema(this.charts.queryId, this.charts.queryRelativeUrl, '', organizationId).subscribe((/**
             * @param {?} value
             * @return {?}
             */
            value => {
                if (value) {
                    if (value.schemaValue.chartsType.series.type && value.schemaValue.chartsType.series.type !== -1) {
                        op = Object.assign({}, value.schemaValue.chartsData, value.schemaValue.chartsStyle);
                        value.schemaValue.chartsType.series.type = value.schemaValue.chartsType.series.type.replace('1', '');
                        op.series.type = value.schemaValue.chartsType.series.type;
                        if (op.series.type !== 'line1') {
                            delete op.series.areaStyle;
                        }
                        else {
                            op.series.areaStyle = {};
                        }
                    }
                }
                /** @type {?} */
                const dynamiccols = [];
                this.bindingList.properties.forEach((/**
                 * @param {?} col
                 * @return {?}
                 */
                col => {
                    /** @type {?} */
                    const dynamiccol = {};
                    dynamiccol.name = col.name;
                    dynamiccol.bindField = col.name;
                    dynamiccol.colWidth = 150;
                    dynamiccol.align = 0;
                    dynamiccol.visible = true;
                    dynamiccol.isFixed = false;
                    dynamiccol.formatType = 0;
                    dynamiccol.formattor = '';
                    dynamiccol.sort = 0;
                    dynamiccol.pageAggreText = null;
                    dynamiccol.pageAggre = 0;
                    dynamiccol.isGroup = false;
                    dynamiccol.groupAggreText = '';
                    dynamiccol.groupAggre = 0;
                    dynamiccol.totalAggreText = '';
                    dynamiccol.totalAggre = 0;
                    dynamiccol.level = 1;
                    dynamiccol.childList = [];
                    dynamiccol.dimension = 0;
                    dynamiccol.rowColor = null;
                    dynamiccols.push(dynamiccol);
                }));
                this.schemaManager.setColInfo(dynamiccols);
                if (op && op.series && op.series.type) {
                    op.series.type = op.series.type.replace('1', '');
                }
                this.charts.setData(this.bindingList.toJSON(), op);
            }));
        }
    }
    /**
     * @private
     * @return {?}
     */
    get bindingList() {
        if (!this.bindingData || !this.viewModel) {
            return undefined;
        }
        // 根实体
        if (this.viewModel.bindingPath === '/' || !this.viewModel.bindingPath) {
            return this.bindingData.list;
        }
        // 子实体
        /** @type {?} */
        let bindingPath = this.viewModel.bindingPath.substr(1);
        bindingPath = bindingPath[0].toLowerCase() + bindingPath.substring(1, bindingPath.length);
        /** @type {?} */
        const paths = bindingPath.split('/');
        /** @type {?} */
        const filteredPaths = paths.filter((/**
         * @param {?} part
         * @return {?}
         */
        (part) => {
            return part !== '';
        }));
        return this.bindingData.getValue(filteredPaths);
    }
}
FormBindingDataDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'formBindingData, [formBindingData]',
            },] }
];
/** @nocollapse */
FormBindingDataDirective.ctorParameters = () => [
    { type: NgxEchartsDirective, decorators: [{ type: Optional }] },
    { type: BindingData, decorators: [{ type: Optional }] },
    { type: ViewModel, decorators: [{ type: Optional }] },
    { type: SchemaManagerService },
    { type: Injector, decorators: [{ type: Optional }] }
];
FormBindingDataDirective.propDecorators = {
    delay: [{ type: Input }],
    interval: [{ type: Input }],
    queryId: [{ type: Input }],
    controlType: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FormBindingDataDirective.prototype.delay;
    /** @type {?} */
    FormBindingDataDirective.prototype.interval;
    /** @type {?} */
    FormBindingDataDirective.prototype.queryId;
    /** @type {?} */
    FormBindingDataDirective.prototype.controlType;
    /**
     * @type {?}
     * @private
     */
    FormBindingDataDirective.prototype.myVar;
    /** @type {?} */
    FormBindingDataDirective.prototype.cache;
    /** @type {?} */
    FormBindingDataDirective.prototype.charts;
    /** @type {?} */
    FormBindingDataDirective.prototype.bindingData;
    /** @type {?} */
    FormBindingDataDirective.prototype.viewModel;
    /**
     * @type {?}
     * @private
     */
    FormBindingDataDirective.prototype.schemaManager;
    /**
     * @type {?}
     * @private
     */
    FormBindingDataDirective.prototype.injector;
}
