// @farris/extend-file-upload/esm2015/lib/components/view-image/f-file-view-image.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/view-image/f-file-view-image.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, Input, ViewChild, ElementRef } from '@angular/core';
export class FFileViewImageComponent {
    constructor() {
        this.maxHeight = 200;
        this.maxWidth = 300;
        this.cls = true;
        this.selfHeight = '50px';
    }
    /**
     * @param {?} fileContent
     * @return {?}
     */
    set file(fileContent) {
        this.tFile = fileContent;
        this.createImageByCanvas();
    }
    /**
     * @return {?}
     */
    get file() {
        return this.tFile;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @private
     * @param {?} originWidth
     * @param {?} originHeight
     * @return {?}
     */
    changeImageSize(originWidth, originHeight) {
        // 图片原始尺寸
        // 图片原始尺寸
        // 目标尺寸
        /** @type {?} */
        var targetWidth = originWidth;
        /** @type {?} */
        var targetHeight = originHeight;
        // 图片尺寸超过400x400的限制
        if (originWidth > this.maxWidth || originHeight > this.maxHeight) {
            if (originWidth / originHeight > this.maxWidth / this.maxHeight) {
                // 更宽，按照宽度限定尺寸
                targetWidth = this.maxWidth;
                targetHeight = Math.round(this.maxWidth * (originHeight / originWidth));
            }
            else {
                targetHeight = this.maxHeight;
                targetWidth = Math.round(this.maxHeight * (originWidth / originHeight));
            }
        }
        this.selfHeight = targetHeight + "px";
        return {
            width: targetWidth,
            height: targetHeight
        };
    }
    // 压缩图片需要的一些元素和对象
    /**
     * @private
     * @return {?}
     */
    createImageByCanvas() {
        /** @type {?} */
        let reader = new FileReader();
        /** @type {?} */
        let tImage = new Image();
        /** @type {?} */
        let canvas = this.canvas.nativeElement;
        /** @type {?} */
        let context = canvas.getContext('2d');
        // base64地址图片加载完毕后
        tImage.onload = (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            let imgSize = this.changeImageSize(tImage.width, tImage.height);
            // canvas对图片进行缩放
            canvas.width = imgSize.width;
            canvas.height = imgSize.height;
            // 清除画布
            context.clearRect(0, 0, this.maxWidth, this.maxHeight);
            // 图片压缩
            context.drawImage(tImage, 0, 0, imgSize.width, imgSize.height);
        });
        // 文件base64化，以便获知图片原始尺寸
        reader.onload = (/**
         * @param {?} ev
         * @return {?}
         */
        (ev) => {
            tImage.src = (/** @type {?} */ (reader.result));
        });
        reader.onerror = (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
        });
        reader.readAsDataURL(this.file);
    }
}
FFileViewImageComponent.decorators = [
    { type: Component, args: [{
                selector: 'f-file-view-image',
                template: "<canvas #canvas class=\"ffileupload--canvas\"></canvas>\r\n",
                styles: [".ffileupload--canvas{margin:0 auto}"]
            }] }
];
/** @nocollapse */
FFileViewImageComponent.ctorParameters = () => [];
FFileViewImageComponent.propDecorators = {
    maxHeight: [{ type: Input }],
    maxWidth: [{ type: Input }],
    cls: [{ type: HostBinding, args: ['class.f-file-view-image',] }],
    selfHeight: [{ type: HostBinding, args: ['style.height',] }],
    canvas: [{ type: ViewChild, args: ['canvas',] }],
    file: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FFileViewImageComponent.prototype.maxHeight;
    /** @type {?} */
    FFileViewImageComponent.prototype.maxWidth;
    /** @type {?} */
    FFileViewImageComponent.prototype.cls;
    /** @type {?} */
    FFileViewImageComponent.prototype.selfHeight;
    /** @type {?} */
    FFileViewImageComponent.prototype.canvas;
    /**
     * @type {?}
     * @private
     */
    FFileViewImageComponent.prototype.tFile;
}
