// @farris/ui-common/esm2015/lib/service/debug.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { fromEvent, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class DebugService {
    constructor() {
        this.isDebug = false;
        this.destroy$ = new Subject();
        this.debugSub = null;
    }
    /**
     * @return {?}
     */
    destroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    /**
     * @private
     * @param {?} type
     * @param {...?} msg
     * @return {?}
     */
    writeMessage(type, ...msg) {
        if (this.isDebug) {
            console[type](...msg);
        }
    }
    /**
     * @param {...?} msg
     * @return {?}
     */
    log(...msg) {
        this.writeMessage('log', msg);
    }
    /**
     * @param {...?} msg
     * @return {?}
     */
    warn(...msg) {
        this.writeMessage('warn', msg);
    }
    /**
     * @param {...?} msg
     * @return {?}
     */
    error(...msg) {
        this.writeMessage('error', msg);
    }
    /**
     * @param {...?} msg
     * @return {?}
     */
    info(...msg) {
        this.writeMessage('info', msg);
    }
    /**
     * @return {?}
     */
    useDebugMode() {
        if (!this.debugSub) {
            this.debugSub = fromEvent(document.body, 'keydown').pipe(takeUntil(this.destroy$)).subscribe((/**
             * @param {?} e
             * @return {?}
             */
            (e) => {
                if (e.ctrlKey && e.shiftKey && e.key === 'D') {
                    e.preventDefault();
                    e.stopPropagation();
                    this.isDebug = !this.isDebug;
                    // console.log(`%c🪲 Farris debug mode is ${ this.isDebug ? 'startting' : 'stopped'}.🪲`, 'color: blue; font-weight:bold;font-size:16px')
                    console.log(`%c🪲 Farris debug mode is ${this.isDebug ? 'startting' : 'stopped'}.🪲`, `color: ${this.isDebug ? 'green' : 'red'}; font-weight:bold;font-size:16px`);
                }
            }));
        }
        return this.debugSub;
    }
}
DebugService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
DebugService.ctorParameters = () => [];
/** @nocollapse */ DebugService.ngInjectableDef = i0.defineInjectable({ factory: function DebugService_Factory() { return new DebugService(); }, token: DebugService, providedIn: "root" });
if (false) {
    /** @type {?} */
    DebugService.prototype.isDebug;
    /**
     * @type {?}
     * @private
     */
    DebugService.prototype.destroy$;
    /**
     * @type {?}
     * @private
     */
    DebugService.prototype.debugSub;
}
