// @farris/ide-devkit/esm5/lib/components/combo-list/combo.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/** @type {?} */
export var DEFAULT_COMBOSTATE = new InjectionToken('init combo state.');
/** @type {?} */
export var COMBO_FACADE_SERVICE = new InjectionToken('combo faceade Service');
/**
 * @record
 */
export function ComboPanelState() { }
if (false) {
    /** @type {?} */
    ComboPanelState.prototype.width;
    /** @type {?} */
    ComboPanelState.prototype.height;
    /** @type {?} */
    ComboPanelState.prototype.left;
    /** @type {?} */
    ComboPanelState.prototype.top;
    /** @type {?} */
    ComboPanelState.prototype.align;
    /** @type {?} */
    ComboPanelState.prototype.open;
}
/**
 * @record
 */
export function ComboState() { }
if (false) {
    /** @type {?} */
    ComboState.prototype.readonly;
    /** @type {?} */
    ComboState.prototype.editable;
    /** @type {?} */
    ComboState.prototype.panel;
}
/** @type {?} */
export var defaultComboState = {
    idField: '',
    textField: '',
    valueField: '',
    disabled: false,
    data: [],
    selections: [],
    readonly: false,
    editable: true,
    multiSelect: false,
    panel: {
        left: 0,
        top: 0,
        width: 300,
        height: 200,
        align: 'auto',
        open: false
    }
};
/**
 * @record
 */
export function SelectItem() { }
if (false) {
    /** @type {?} */
    SelectItem.prototype.index;
    /** @type {?} */
    SelectItem.prototype.data;
}
/**
 * @record
 */
export function RemoteParams() { }
if (false) {
    /** @type {?|undefined} */
    RemoteParams.prototype.pageIndex;
    /** @type {?|undefined} */
    RemoteParams.prototype.pageSize;
    /** @type {?|undefined} */
    RemoteParams.prototype.condition;
    /** @type {?|undefined} */
    RemoteParams.prototype.helpId;
    /** @type {?|undefined} */
    RemoteParams.prototype.searchField;
    /** @type {?|undefined} */
    RemoteParams.prototype.searchValue;
    /** @type {?|undefined} */
    RemoteParams.prototype.sortName;
    /** @type {?|undefined} */
    RemoteParams.prototype.sortOrder;
}
/**
 * @record
 */
export function NavigationFilter() { }
if (false) {
    /** @type {?|undefined} */
    NavigationFilter.prototype.searchField;
    /** @type {?|undefined} */
    NavigationFilter.prototype.searchValue;
    /** @type {?|undefined} */
    NavigationFilter.prototype.selected;
    /** @type {?|undefined} */
    NavigationFilter.prototype.idValue;
}
/**
 * @record
 */
export function SearchParam() { }
if (false) {
    /** @type {?|undefined} */
    SearchParam.prototype.searchField;
    /** @type {?|undefined} */
    SearchParam.prototype.searchValue;
    /** @type {?|undefined} */
    SearchParam.prototype.category;
    /** @type {?|undefined} */
    SearchParam.prototype.navigation;
}
