// @gsp-wf/wf-bizprocess-lookup/bundles/gsp-wf-wf-bizprocess-lookup.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('rxjs'), require('rxjs/operators'), require('@angular/core'), require('@angular/common'), require('@angular/forms'), require('@angular/common/http'), require('@farris/ui-lookup'), require('@farris/ui-modal'), require('@farris/ui-messager'), require('@ecp-caf/caf-common')) :
    typeof define === 'function' && define.amd ? define('@gsp-wf/wf-bizprocess-lookup', ['exports', 'rxjs', 'rxjs/operators', '@angular/core', '@angular/common', '@angular/forms', '@angular/common/http', '@farris/ui-lookup', '@farris/ui-modal', '@farris/ui-messager', '@ecp-caf/caf-common'], factory) :
    (factory((global['gsp-wf'] = global['gsp-wf'] || {}, global['gsp-wf']['wf-bizprocess-lookup'] = {}),global.rxjs,global.rxjs.operators,global.ng.core,global.ng.common,global.ng.forms,global.ng.common.http,global.uiLookup,global.uiModal,global.uiMessager,global.cafCommon));
}(this, (function (exports,rxjs,operators,i0,common,forms,http,uiLookup,uiModal,uiMessager,cafCommon) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var WfBizprocessLookupService = /** @class */ (function () {
        function WfBizprocessLookupService() {
        }
        WfBizprocessLookupService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */
        WfBizprocessLookupService.ctorParameters = function () { return []; };
        /** @nocollapse */ WfBizprocessLookupService.ngInjectableDef = i0.defineInjectable({ factory: function WfBizprocessLookupService_Factory() { return new WfBizprocessLookupService(); }, token: WfBizprocessLookupService, providedIn: "root" });
        return WfBizprocessLookupService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var WFBizProcessLookUpUIState = /** @class */ (function () {
        function WFBizProcessLookUpUIState() {
            this.rtProcessUri = '/api/runtime/wf/v1.0/procDefs/processDefinitionKeys/query';
            this.flowFormUri = '/api/runtime/wf/v1.0/flowForm/query';
        }
        WFBizProcessLookUpUIState.decorators = [
            { type: i0.Injectable }
        ];
        /** @nocollapse */
        WFBizProcessLookUpUIState.ctorParameters = function () { return []; };
        return WFBizProcessLookUpUIState;
    }());

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var WFBizProcessLookUpService = /** @class */ (function () {
        function WFBizProcessLookUpService(http$$1) {
            this.http = http$$1;
        }
        /**
         * @param {?} url
         * @param {?=} params
         * @return {?}
         */
        WFBizProcessLookUpService.prototype.getData = /**
         * @param {?} url
         * @param {?=} params
         * @return {?}
         */
            function (url, params) {
                var e_1, _a;
                /** @type {?} */
                var customData = params.customData;
                /** @type {?} */
                var index = params.pageIndex || 1;
                /** @type {?} */
                var size = params.pageSize || 20;
                /** @type {?} */
                var searchValue = JSON.parse(params.searchValue);
                if (searchValue.searchValue || customData.filter) {
                    /** @type {?} */
                    var query = '?param=';
                    /** @type {?} */
                    var filter = '{';
                    if (searchValue.searchValue) {
                        if (searchValue.searchField === '*') {
                            filter += "\"allColumns\":\"" + searchValue.searchValue + "\",";
                        }
                        else {
                            filter += "\"" + searchValue.searchField + "\":\"" + searchValue.searchValue + "\",";
                        }
                    }
                    if (customData.filter) {
                        /** @type {?} */
                        var keys = Object.keys(customData.filter);
                        if (keys && keys.length > 0) {
                            try {
                                for (var keys_1 = __values(keys), keys_1_1 = keys_1.next(); !keys_1_1.done; keys_1_1 = keys_1.next()) {
                                    var key = keys_1_1.value;
                                    if (customData.filter[key]) {
                                        filter += "\"" + key + "\":\"" + customData.filter[key] + "\",";
                                    }
                                }
                            }
                            catch (e_1_1) {
                                e_1 = { error: e_1_1 };
                            }
                            finally {
                                try {
                                    if (keys_1_1 && !keys_1_1.done && (_a = keys_1.return))
                                        _a.call(keys_1);
                                }
                                finally {
                                    if (e_1)
                                        throw e_1.error;
                                }
                            }
                        }
                    }
                    if (filter.substr(filter.length - 1, 1) === ',') {
                        filter = filter.substring(0, filter.length - 1);
                    }
                    filter += '}';
                    filter = encodeURIComponent(filter);
                    query += (filter + ("&pageNum=" + index + "&pageSize=" + size));
                    url += query;
                }
                else {
                    url += "?pageNum=" + index + "&pageSize=" + size;
                }
                return this.http.get(url).pipe(operators.map(( /**
                 * @param {?} re
                 * @return {?}
                 */function (re) {
                    return ({
                        items: re.content,
                        total: re.totalElements,
                        pageInfo: {
                            pageIndex: re.pageable.pageNumber + 1,
                            pageSize: re.pageable.pageSize,
                            pageList: [20, 40, 60, 80, 100],
                            enablePager: true
                        }
                    });
                })));
            };
        WFBizProcessLookUpService.decorators = [
            { type: i0.Injectable }
        ];
        /** @nocollapse */
        WFBizProcessLookUpService.ctorParameters = function () {
            return [
                { type: cafCommon.HttpService }
            ];
        };
        return WFBizProcessLookUpService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var WfBizprocessLookupComponent = /** @class */ (function () {
        function WfBizprocessLookupComponent(uiState) {
            var _this = this;
            this.uiState = uiState;
            /**
             * 宽度
             */
            this.width = 550;
            /**
             * 取数源
             */
            this.sourceType = 'rtProcess';
            /**
             * 只读
             */
            this.disabled = false;
            /**
             * 是否单选
             */
            this.singleSelect = true;
            /**
             * 是否分页，默认true
             */
            this.pagination = true;
            /**
             * 首次加载第几页，默认1
             */
            this.pageIndex = 1;
            /**
             * 每页加载数，默认20
             */
            this.pageSize = 20;
            // @Input() beforeClose: (selectData: any) => Observable<PickedResult>;
            /**
             * 帮助选择后事件
             */
            this.afterConfirm = new i0.EventEmitter();
            this.dialogClosed = new i0.EventEmitter();
            this.clear = new i0.EventEmitter();
            this.columns = [];
            /**
             * 帮助前事件,强行传递过滤条件
             */
            this.beforeOpen = ( /**
             * @param {?} arg
             * @return {?}
             */function (arg) {
                // if (!this.filter) {
                //   // 可能调用方不传递，识别为非空
                //   this.filter = {};
                // }
                // this.filter.pageIndex = this.pageIndex;
                // this.filter.pageSize = this.pageSize;
                // 把过滤数据和当前的数据传递到service层
                return rxjs.of({
                    showDialog: true, data: {
                        filter: _this.filter
                    }
                });
            });
            this.dictPicked = ( /**
             * @param {?} arg
             * @return {?}
             */function (arg) {
                _this.afterConfirm.next(arg);
                return rxjs.of({ closeDialog: true });
            });
        }
        /**
         * @return {?}
         */
        WfBizprocessLookupComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                switch (this.sourceType) {
                    case 'rtProcess':
                        this.columns = [
                            { title: 'Key值', field: 'id', width: '80', searchField: 'id', },
                            { title: '名称', field: 'name', width: '100', searchField: 'name', }
                        ];
                        this.dataUri = this.uiState.rtProcessUri;
                        break;
                    case 'flowform':
                        this.columns = [
                            { title: '编号', field: 'code', width: '80', searchField: 'code', },
                            { title: '名称', field: 'name', width: '100', searchField: 'name', }
                        ];
                        this.dataUri = this.uiState.flowFormUri;
                        break;
                    default:
                        break;
                }
            };
        /**
         * @param {?} arg
         * @return {?}
         */
        WfBizprocessLookupComponent.prototype.afterClear = /**
         * @param {?} arg
         * @return {?}
         */
            function (arg) {
                this.clear.next(arg);
            };
        /** 手动显示帮助 */
        /**
         * 手动显示帮助
         * @return {?}
         */
        WfBizprocessLookupComponent.prototype.showHelp = /**
         * 手动显示帮助
         * @return {?}
         */
            function () {
                this.lookUp.showDialog();
            };
        WfBizprocessLookupComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'wf-bizprocess-lookup',
                        template: "\n    <farris-lookup-grid #lookup\n      [disabled]=\"disabled\"\n      [editable]=\"false\"\n      [bindingData]=\"bindingData\"\n      [(ngModel)]=\"displayTxt\"\n      [columns]=\"columns\"\n      [singleSelect]=\"singleSelect\"\n      [title]=\"title\"\n      [displayType]=\"'List'\"\n      [uri]=\"dataUri\"\n      [idField]=\"'id'\"\n      [pageIndex]=\"pageIndex\"\n      [pageSize]=\"pageSize\"\n      [pagination]=\"pagination\"\n      [valueField]=\"'id'\"\n      [textField]=\"'name'\"\n      (clear)=\"afterClear($event)\"\n      [dialogWidth]=\"width\"\n      [dictPicking]=\"beforeOpen\"\n      [dictPicked]=\"dictPicked\">\n    </farris-lookup-grid>\n  ",
                        providers: [{ provide: uiLookup.ServerSideToken, useClass: WFBizProcessLookUpService }]
                    }] }
        ];
        /** @nocollapse */
        WfBizprocessLookupComponent.ctorParameters = function () {
            return [
                { type: WFBizProcessLookUpUIState }
            ];
        };
        WfBizprocessLookupComponent.propDecorators = {
            title: [{ type: i0.Input }],
            width: [{ type: i0.Input }],
            sourceType: [{ type: i0.Input }],
            disabled: [{ type: i0.Input }],
            singleSelect: [{ type: i0.Input }],
            pagination: [{ type: i0.Input }],
            pageIndex: [{ type: i0.Input }],
            pageSize: [{ type: i0.Input }],
            displayTxt: [{ type: i0.Input }],
            bindingData: [{ type: i0.Input }],
            filter: [{ type: i0.Input }],
            afterConfirm: [{ type: i0.Output }],
            dialogClosed: [{ type: i0.Output }],
            clear: [{ type: i0.Output }],
            lookUp: [{ type: i0.ViewChild, args: ['lookup',] }]
        };
        return WfBizprocessLookupComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var WfBizprocessLookupModule = /** @class */ (function () {
        function WfBizprocessLookupModule() {
        }
        WfBizprocessLookupModule.decorators = [
            { type: i0.NgModule, args: [{
                        declarations: [WfBizprocessLookupComponent],
                        imports: [
                            common.CommonModule,
                            forms.FormsModule,
                            http.HttpClientModule,
                            uiLookup.LookupModule,
                            uiModal.ModalModule.forRoot(),
                            uiMessager.MessagerModule.forRoot(),
                        ],
                        providers: [
                            cafCommon.HttpService,
                            WFBizProcessLookUpUIState
                        ],
                        exports: [WfBizprocessLookupComponent]
                    },] }
        ];
        return WfBizprocessLookupModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.WfBizprocessLookupService = WfBizprocessLookupService;
    exports.WfBizprocessLookupComponent = WfBizprocessLookupComponent;
    exports.WfBizprocessLookupModule = WfBizprocessLookupModule;
    exports.WFBizProcessLookUpService = WFBizProcessLookUpService;
    exports.WFBizProcessLookUpUIState = WFBizProcessLookUpUIState;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
