// @farris/ui-datagrid-filter/esm5/lib/operations/operators.js
var _a;
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 服务器端支持比较操作
 *  Equal(0),
    NotEqual(1),
    Greater(2),
    GreaterOrEqual(3),
    Less(4),
    LessOrEqual(5),
    Like(6),
    LikeStartWith(7),
    LikeEndWith(8),
    NotLike(9),
    NotLikeStartWith(10),
    NotLikeEndWith(11),
    Is(12),
    IsNot(13),
    In(14),
    NotIn(15),
    LikeIgnoreCase(16);
 */
/** @enum {number} */
var FilterOperator = {
    Equal: 0,
    NotEqual: 1,
    Greater: 2,
    GreaterOrEqual: 3,
    Less: 4,
    LessOrEqual: 5,
    Contains: 6,
    NotContains: 9,
    In: 14,
    NotIn: 15,
    Empty: 1001,
    NotEmpty: 1002,
    Null: 1003,
    NotNull: 1004,
};
export { FilterOperator };
FilterOperator[FilterOperator.Equal] = 'Equal';
FilterOperator[FilterOperator.NotEqual] = 'NotEqual';
FilterOperator[FilterOperator.Greater] = 'Greater';
FilterOperator[FilterOperator.GreaterOrEqual] = 'GreaterOrEqual';
FilterOperator[FilterOperator.Less] = 'Less';
FilterOperator[FilterOperator.LessOrEqual] = 'LessOrEqual';
FilterOperator[FilterOperator.Contains] = 'Contains';
FilterOperator[FilterOperator.NotContains] = 'NotContains';
FilterOperator[FilterOperator.In] = 'In';
FilterOperator[FilterOperator.NotIn] = 'NotIn';
FilterOperator[FilterOperator.Empty] = 'Empty';
FilterOperator[FilterOperator.NotEmpty] = 'NotEmpty';
FilterOperator[FilterOperator.Null] = 'Null';
FilterOperator[FilterOperator.NotNull] = 'NotNull';
/** @enum {string} */
var FilterConditionValue = {
    All: '',
    Empty: 'null',
};
export { FilterConditionValue };
/**
 * @record
 */
export function FilterEnumSetting() { }
if (false) {
    /** @type {?|undefined} */
    FilterEnumSetting.prototype.valueField;
    /** @type {?|undefined} */
    FilterEnumSetting.prototype.textField;
    /** @type {?|undefined} */
    FilterEnumSetting.prototype.data;
    /** @type {?|undefined} */
    FilterEnumSetting.prototype.idField;
}
/**
 * @record
 */
export function FilterRowOperator() { }
if (false) {
    /** @type {?} */
    FilterRowOperator.prototype.code;
    /** @type {?} */
    FilterRowOperator.prototype.label;
    /** @type {?} */
    FilterRowOperator.prototype.isDefault;
}
/**
 * @record
 */
export function FilterCondition() { }
if (false) {
    /** @type {?|undefined} */
    FilterCondition.prototype.operator1;
    /** @type {?|undefined} */
    FilterCondition.prototype.value1;
    /** @type {?|undefined} */
    FilterCondition.prototype.relation;
    /** @type {?|undefined} */
    FilterCondition.prototype.operator2;
    /** @type {?|undefined} */
    FilterCondition.prototype.value2;
}
/**
 * @record
 */
export function ColumnFilterCondition() { }
/**
 * @record
 */
export function FilterData() { }
if (false) {
    /** @type {?} */
    FilterData.prototype.filterField;
    /** @type {?} */
    FilterData.prototype.value;
    /** @type {?|undefined} */
    FilterData.prototype.lbracket;
    /** @type {?|undefined} */
    FilterData.prototype.rbracket;
    /** @type {?|undefined} */
    FilterData.prototype.relation;
    /** @type {?|undefined} */
    FilterData.prototype.compare;
}
/** @type {?} */
export var AllFilterOperator = (_a = {},
    _a[FilterOperator.Equal] = '等于',
    _a[FilterOperator.NotEqual] = '不等于',
    _a[FilterOperator.Greater] = '大于',
    _a[FilterOperator.GreaterOrEqual] = '大于等于',
    _a[FilterOperator.Less] = '小于',
    _a[FilterOperator.LessOrEqual] = '小于等于',
    _a[FilterOperator.Contains] = '包含',
    _a[FilterOperator.NotContains] = '不包含',
    _a[FilterOperator.NotEmpty] = '不为空',
    _a[FilterOperator.Empty] = '空',
    _a);
/**
 * 文本操作符
 * @type {?}
 */
export var TextFilterOperators = [
    { code: FilterOperator.Empty, label: AllFilterOperator[FilterOperator.Empty], isDefault: false },
    { code: FilterOperator.NotEmpty, label: AllFilterOperator[FilterOperator.NotEmpty], isDefault: false },
    { code: FilterOperator.Equal, label: AllFilterOperator[FilterOperator.Equal], isDefault: false },
    { code: FilterOperator.NotEqual, label: AllFilterOperator[FilterOperator.NotEqual], isDefault: false },
    { code: FilterOperator.Contains, label: AllFilterOperator[FilterOperator.Contains], isDefault: true },
    { code: FilterOperator.NotContains, label: AllFilterOperator[FilterOperator.NotContains], isDefault: false }
];
/**
 * 日期或数字操作符
 * @type {?}
 */
export var NumberOrDateFilterOperators = [
    { code: FilterOperator.Empty, label: AllFilterOperator[FilterOperator.Empty], isDefault: false },
    { code: FilterOperator.NotEmpty, label: AllFilterOperator[FilterOperator.NotEmpty], isDefault: false },
    { code: FilterOperator.Equal, label: AllFilterOperator[FilterOperator.Equal], isDefault: false },
    { code: FilterOperator.NotEqual, label: AllFilterOperator[FilterOperator.NotEqual], isDefault: false },
    { code: FilterOperator.Greater, label: AllFilterOperator[FilterOperator.Greater], isDefault: false },
    { code: FilterOperator.GreaterOrEqual, label: AllFilterOperator[FilterOperator.GreaterOrEqual], isDefault: false },
    { code: FilterOperator.Less, label: AllFilterOperator[FilterOperator.Less], isDefault: false },
    { code: FilterOperator.LessOrEqual, label: AllFilterOperator[FilterOperator.LessOrEqual], isDefault: false }
];
