// @gsp-svc/formdoc-upload/esm2015/lib/uploaddialog/uploaddialog/uploaddialog.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Input } from '@angular/core';
import { BsModalService } from '@farris/ui-modal';
import { FileUploadComponent } from '../../upload/file-upload/file-upload.component';
export class UploadDialogComponent {
    /**
     * @param {?} modalService
     */
    constructor(modalService) {
        this.modalService = modalService;
        this.metadataIdList = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    showDialog() {
        this.dlg = this.modalService.show(FileUploadComponent, {
            title: '选择文件',
            width: 800, height: 500,
            buttons: [
                {
                    text: '上传', cls: 'k-button k-button-icontext k-primary', iconCls: "k-icon k-i-check", handle: (/**
                     * @return {?}
                     */
                    () => {
                        this.fileUpload.upload();
                        this.dlg.close();
                    })
                },
                {
                    text: '关闭', cls: 'k-button k-button-icontext', iconCls: "k-icon k-i-close", handle: (/**
                     * @return {?}
                     */
                    () => {
                        this.fileUpload.cancel();
                        this.dlg.close();
                    })
                }
            ],
            showButtons: true,
        });
        this.dlg.content.formId = this.formId;
    }
}
UploadDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-uploaddialog',
                template: "\n",
                styles: [""]
            }] }
];
/** @nocollapse */
UploadDialogComponent.ctorParameters = () => [
    { type: BsModalService }
];
UploadDialogComponent.propDecorators = {
    formId: [{ type: Input }],
    fileUpload: [{ type: ViewChild, args: [FileUploadComponent,] }]
};
if (false) {
    /** @type {?} */
    UploadDialogComponent.prototype.dlg;
    /** @type {?} */
    UploadDialogComponent.prototype.metadataIdList;
    /** @type {?} */
    UploadDialogComponent.prototype.formId;
    /** @type {?} */
    UploadDialogComponent.prototype.fileUpload;
    /**
     * @type {?}
     * @private
     */
    UploadDialogComponent.prototype.modalService;
}
