// @edp-bif/runtime-api/esm5/lib/bill-tracker/config/bill-tracker-graph-options.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/config/bill-tracker-graph-options.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 单据追踪图流程图配置
 */
var /**
 * 单据追踪图流程图配置
 */
BillTrackerGraphOptions = /** @class */ (function () {
    function BillTrackerGraphOptions() {
        /**
         * 节点宽度
         */
        this.nodeWidth = 100;
        /**
         * 节点高度
         */
        this.nodeHeight = 60;
        /**
         * 上查/下查/上折/下折按钮宽度
         */
        this.overlayWidth = 10;
        /**
         * 上查/下查/上折/下折按钮高度
         */
        this.overlayHeight = 10;
        /**
         * 显示直接关联的下游单据
         */
        this.showDownBills = true;
        /**
         * 显示直接关联的上游单据
         */
        this.showUpBills = true;
    }
    return BillTrackerGraphOptions;
}());
/**
 * 单据追踪图流程图配置
 */
export { BillTrackerGraphOptions };
if (false) {
    /**
     * 上查按钮图片
     * @type {?}
     */
    BillTrackerGraphOptions.prototype.upSearchImage;
    /**
     * 下查按钮图片
     * @type {?}
     */
    BillTrackerGraphOptions.prototype.downSearchImage;
    /**
     * 上折按钮图片
     * @type {?}
     */
    BillTrackerGraphOptions.prototype.upFolderImage;
    /**
     * 下折按钮图片
     * @type {?}
     */
    BillTrackerGraphOptions.prototype.downFolderImage;
    /**
     * 节点宽度
     * @type {?}
     */
    BillTrackerGraphOptions.prototype.nodeWidth;
    /**
     * 节点高度
     * @type {?}
     */
    BillTrackerGraphOptions.prototype.nodeHeight;
    /**
     * 上查/下查/上折/下折按钮宽度
     * @type {?}
     */
    BillTrackerGraphOptions.prototype.overlayWidth;
    /**
     * 上查/下查/上折/下折按钮高度
     * @type {?}
     */
    BillTrackerGraphOptions.prototype.overlayHeight;
    /**
     * 显示直接关联的下游单据
     * @type {?}
     */
    BillTrackerGraphOptions.prototype.showDownBills;
    /**
     * 显示直接关联的上游单据
     * @type {?}
     */
    BillTrackerGraphOptions.prototype.showUpBills;
    /* Skipping unhandled member: [key: string]: any;*/
}
