// @farris/ide-devkit/esm2015/lib/components/panel/sash.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, Output, ViewContainerRef, HostListener, Renderer2 } from '@angular/core';
export class SashDirective {
    /**
     * @param {?} _elementRef
     * @param {?} renderer
     * @param {?} _viewContainerRef
     */
    constructor(_elementRef, renderer, _viewContainerRef) {
        this._elementRef = _elementRef;
        this.renderer = renderer;
        this._viewContainerRef = _viewContainerRef;
        this.widthChanged = new EventEmitter();
    }
    /**
     * @return {?}
     */
    onMouseDown() {
        console.log('mousedown');
        this.bindDocumentClickListener();
    }
    /**
     * @return {?}
     */
    onMouseUp() {
        console.log('mouseup');
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
    }
    /**
     * @return {?}
     */
    bindDocumentClickListener() {
        if (!this.documentClickListener) {
            this.documentClickListener = this.renderer.listen('document', 'mouseup', (event) => {
                console.log('mouseup');
                console.log(event);
                this.unbindDocumentMoveListener();
                this.unbindDocumentClickListener();
            });
            if (!this.documentMoveListener) {
                this.bindDocumentClickListener = this.renderer.listen('document', 'mousemove', (event) => {
                    this.panelWidth = event.clientX;
                    /** @type {?} */
                    const newWidth = (/** @type {?} */ (event.clientX));
                    this.widthChanged.emit(newWidth);
                });
            }
        }
    }
    /**
     * @return {?}
     */
    unbindDocumentClickListener() {
        if (this.documentClickListener) {
            this.documentClickListener();
            this.documentClickListener = null;
        }
    }
    /**
     * @return {?}
     */
    unbindDocumentMoveListener() {
        if (this.bindDocumentClickListener) {
            this.bindDocumentClickListener();
            this.bindDocumentClickListener = null;
        }
    }
}
SashDirective.decorators = [
    { type: Directive, args: [{
                selector: '[sash]',
                exportAs: 'sash'
            },] }
];
/** @nocollapse */
SashDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: ViewContainerRef }
];
SashDirective.propDecorators = {
    panelWidth: [{ type: Input }],
    widthChanged: [{ type: Output }],
    onMouseDown: [{ type: HostListener, args: ['mousedown',] }],
    onMouseUp: [{ type: HostListener, args: ['mouseup',] }]
};
if (false) {
    /** @type {?} */
    SashDirective.prototype.documentClickListener;
    /** @type {?} */
    SashDirective.prototype.documentMoveListener;
    /** @type {?} */
    SashDirective.prototype.panelWidth;
    /** @type {?} */
    SashDirective.prototype.widthChanged;
    /**
     * @type {?}
     * @private
     */
    SashDirective.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    SashDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    SashDirective.prototype._viewContainerRef;
}
