// @gspwidget/widget-devkit/esm5/lib/preset/preset.index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 存放所有从预制元部件中抽取出来的公共值、方法。
 * 好处是在此库中统一维护、管理，不会被打到每个元部件中。
 */
export { JumpTypes, jumpTypeOption, funcOption, dataJumpOption, datalessJumpOption, CommonPropertyOptions } from './preset.jump';
export { filterYData, echartsGradient, tempLabel, EchartsOptions, chartsConsts, getChartColorsPure, getChartColorsGradient, getChartColorsGradientOne, EchartsPropertyOptions, tempMap } from './preset.echarts';
/**
 * dataItem 这个名字会影响 exp 中的写法，不可随意更改。
 * @param {?} html
 * @param {?} dataItem
 * @return {?}
 */
export function resolveTemplate(html, dataItem) {
    /** @type {?} */
    var expPatt = /{{[\d\D]+?}}/gm;
    if (!html)
        return '';
    /** @type {?} */
    var exps = html.match(expPatt) || [];
    try {
        exps.forEach((/**
         * @param {?} exp
         * @return {?}
         */
        function (exp) {
            html = html.replace(exp, eval(exp.slice(2, -2)));
        }));
    }
    catch (err) {
        return html;
    }
    return html;
}
