// @gspwidget/widget-devkit/esm5/lib/preset/zrender.gradient.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function GradientObject() { }
if (false) {
    /** @type {?|undefined} */
    GradientObject.prototype.id;
    /** @type {?} */
    GradientObject.prototype.type;
    /** @type {?} */
    GradientObject.prototype.colorStops;
}
/**
 * @record
 */
export function InnerGradientObject() { }
if (false) {
    /** @type {?} */
    InnerGradientObject.prototype.__canvasGradient;
}
/**
 * @record
 */
export function GradientColorStop() { }
if (false) {
    /** @type {?} */
    GradientColorStop.prototype.offset;
    /** @type {?} */
    GradientColorStop.prototype.color;
}
var Gradient = /** @class */ (function () {
    function Gradient(colorStops) {
        this.colorStops = colorStops || [];
    }
    /**
     * @param {?} offset
     * @param {?} color
     * @return {?}
     */
    Gradient.prototype.addColorStop = /**
     * @param {?} offset
     * @param {?} color
     * @return {?}
     */
    function (offset, color) {
        this.colorStops.push({
            offset: offset,
            color: color,
        });
    };
    return Gradient;
}());
export default Gradient;
if (false) {
    /** @type {?} */
    Gradient.prototype.id;
    /** @type {?} */
    Gradient.prototype.type;
    /** @type {?} */
    Gradient.prototype.colorStops;
}
