// @farris/ui-list-view/esm2015/lib/template.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, TemplateRef } from '@angular/core';
export class ListViewHeaderTemplateDirective {
    /**
     * @param {?} template
     */
    constructor(template) {
        this.template = template;
    }
}
ListViewHeaderTemplateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[listViewHeader]'
            },] }
];
/** @nocollapse */
ListViewHeaderTemplateDirective.ctorParameters = () => [
    { type: TemplateRef }
];
if (false) {
    /** @type {?} */
    ListViewHeaderTemplateDirective.prototype.template;
}
export class ListViewTemplateDirective {
    /**
     * @param {?} template
     */
    constructor(template) {
        this.template = template;
    }
}
ListViewTemplateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[listTemplate]'
            },] }
];
/** @nocollapse */
ListViewTemplateDirective.ctorParameters = () => [
    { type: TemplateRef }
];
if (false) {
    /** @type {?} */
    ListViewTemplateDirective.prototype.template;
}
export class ListViewFooterTemplateDirective {
    /**
     * @param {?} template
     */
    constructor(template) {
        this.template = template;
    }
}
ListViewFooterTemplateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[listViewFooter]'
            },] }
];
/** @nocollapse */
ListViewFooterTemplateDirective.ctorParameters = () => [
    { type: TemplateRef }
];
if (false) {
    /** @type {?} */
    ListViewFooterTemplateDirective.prototype.template;
}
export class ListViewEmptyTemplateDirective {
    /**
     * @param {?} template
     */
    constructor(template) {
        this.template = template;
    }
}
ListViewEmptyTemplateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[listViewEmpty]'
            },] }
];
/** @nocollapse */
ListViewEmptyTemplateDirective.ctorParameters = () => [
    { type: TemplateRef }
];
if (false) {
    /** @type {?} */
    ListViewEmptyTemplateDirective.prototype.template;
}
