// @farris/ui-common/esm5/lib/directives/farris-template.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef } from '@angular/core';
var FarrisTemplateDirective = /** @class */ (function () {
    function FarrisTemplateDirective(template) {
        this.template = template;
    }
    /**
     * @return {?}
     */
    FarrisTemplateDirective.prototype.getType = /**
     * @return {?}
     */
    function () {
        return this.name;
    };
    FarrisTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farrisTemplate]'
                },] }
    ];
    /** @nocollapse */
    FarrisTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    FarrisTemplateDirective.propDecorators = {
        type: [{ type: Input }],
        name: [{ type: Input, args: ['farrisTemplate',] }]
    };
    return FarrisTemplateDirective;
}());
export { FarrisTemplateDirective };
if (false) {
    /** @type {?} */
    FarrisTemplateDirective.prototype.type;
    /** @type {?} */
    FarrisTemplateDirective.prototype.name;
    /** @type {?} */
    FarrisTemplateDirective.prototype.template;
}
