// @gsp-wf/wf-process-editor/esm5/lib/designer/WfGraph.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Graph } from "@edp-pmf/grapheditor";
import { FlowNodeState } from "@edp-pmf/bpmn-model";
import { mxConstants, mxUtils } from "@edp-pmf/mxgraph-ts";
import { WfProcessCell } from "./wf-cell/WfProcessCell";
/**
 * 工作流画布类
 */
var /**
 * 工作流画布类
 */
WfGraph = /** @class */ (function (_super) {
    tslib_1.__extends(WfGraph, _super);
    function WfGraph(editorUi, container, model, renderHint, stylesheet, themes) {
        var _this = _super.call(this, editorUi, container, model, renderHint, stylesheet, themes) || this;
        _this.localePipe = null;
        /**
         * 流程元素实例
         */
        _this.processCell = new WfProcessCell();
        return _this;
    }
    /**
     * @param {?} cells
     * @param {?} sender
     * @param {?} eventObject
     * @return {?}
     */
    WfGraph.prototype.bpmnCellsSelectionChanged = /**
     * @param {?} cells
     * @param {?} sender
     * @param {?} eventObject
     * @return {?}
     */
    function (cells, sender, eventObject) {
        var _this = this;
        _super.prototype.bpmnCellsSelectionChanged.call(this, cells, sender, eventObject);
        // 修改选中节点样式
        /** @type {?} */
        var selected = (/** @type {?} */ (this.getSelectionCells()));
        /** @type {?} */
        var all = (/** @type {?} */ (this.getChildVertices(this.getDefaultParent())));
        all.forEach((/**
         * @param {?} c
         * @return {?}
         */
        function (c) {
            if (selected.includes(c) && c.imgUrls && c.imgUrls[FlowNodeState.Selected]) {
                c.setStyle(mxUtils.setStyle(c.getStyle(), mxConstants.STYLE_IMAGE, c.imgUrls[FlowNodeState.Selected]));
            }
            else {
                c.setStyle(mxUtils.setStyle(c.getStyle(), mxConstants.STYLE_IMAGE, c.imgUrls[FlowNodeState.Default]));
            }
            _this.refresh(c);
        }));
    };
    /**
     * 重写创建流程元素实例方法
     * @param bpmnModel
     * @returns
     */
    /**
     * 重写创建流程元素实例方法
     * @param {?} bpmnModel
     * @return {?}
     */
    WfGraph.prototype.createProcessCell = /**
     * 重写创建流程元素实例方法
     * @param {?} bpmnModel
     * @return {?}
     */
    function (bpmnModel) {
        /** @type {?} */
        var processCell = new WfProcessCell();
        processCell.process = bpmnModel.DefaultProcess;
        processCell.plane = bpmnModel.DefaultDiagram.BpmnPlane;
        processCell.editorUi = this.editorUi;
        return processCell;
    };
    /**
     * 给定新流程信息创建BpmnModel
     * @param newProcInfo
     * @param bizCmps
     */
    /**
     * 给定新流程信息创建BpmnModel
     * @param {?} newProcInfo
     * @param {?} bizCmps
     * @return {?}
     */
    WfGraph.prototype.createBpmnModelWithProcInfo = /**
     * 给定新流程信息创建BpmnModel
     * @param {?} newProcInfo
     * @param {?} bizCmps
     * @return {?}
     */
    function (newProcInfo, bizCmps) {
        var _a;
        /** @type {?} */
        var bpmnModel = this.createBpmnModel();
        this.bpmnModel = bpmnModel;
        this.processCell = this.createProcessCell(this.bpmnModel);
        if (this.editorUi.langCode === 'zh-CHS') {
            this.bpmnModel.DefaultProcess.Name = newProcInfo.name || bpmnModel.DefaultProcess.Name;
        }
        this.bpmnModel.DefaultProcess.NameLanguage = (_a = {},
            _a[this.editorUi.langCode] = newProcInfo.name || bpmnModel.DefaultProcess.Name,
            _a);
        //设置流程Header(业务对象ID、流程分类ID)
        this.processCell.setProcessHeader(newProcInfo.ffId, newProcInfo.bizObjId, newProcInfo.ffName, (/** @type {?} */ (newProcInfo.startMode)));
        // 基于流程分类绑定参数，流程构件，流程表单
        this.processCell.updateFlowFormInfo(newProcInfo.formalParameters, newProcInfo.processBizComponnets, newProcInfo.executeForms, newProcInfo.customizeActions, bizCmps);
        this.processCell.modifyProcessTitleExpression(newProcInfo.processDescription);
    };
    /**
     * @return {?}
     */
    WfGraph.prototype.setDefaultStyles = /**
     * @return {?}
     */
    function () {
        _super.prototype.setDefaultStyles.call(this);
        /** @type {?} */
        var styleSheet = this.getStylesheet();
        // 连线样式默认为正交
        /** @type {?} */
        var edgeStyle = styleSheet.getDefaultEdgeStyle();
        edgeStyle[mxConstants.STYLE_ROUNDED] = 1;
    };
    return WfGraph;
}(Graph));
/**
 * 工作流画布类
 */
export { WfGraph };
if (false) {
    /** @type {?} */
    WfGraph.prototype.localePipe;
    /** @type {?} */
    WfGraph.prototype.editorUi;
    /**
     * 流程元素实例
     * @type {?}
     */
    WfGraph.prototype.processCell;
}
