// @farris/kendo-binding/esm2015/lib/directives/localization/farris_number_localization.directive.js
import { Directive, Inject, Injector } from "@angular/core";
import { UserSettingsToken } from "@farris/devkit";
import { NumberSpinnerComponent } from '@farris/ui-number-spinner';
export class FarrisNumberLocalizationDirective {
    constructor(injector, userSettings, componentRef) {
        this.injector = injector;
        this.userSettings = userSettings;
        this.componentRef = componentRef;
    }
    ngOnInit() {
        this.localize();
    }
    localize() {
        return;
        // 涉及金额等问题，数字格式化不予处理
        if (this.userSettings && this.userSettings.numberFormat && this.componentRef) {
            const { negativeSign = null, numberDecimalDigits = null, numberDecimalSeparator = null, numberGroupSeparator = null } = this.userSettings.numberFormat || {};
            // 负号
            if (negativeSign) {
                // this.componentRef.prefix = negativeSign;
            }
            // 小数位数
            if (numberDecimalDigits !== null) {
                this.componentRef.precision = numberDecimalDigits;
                this.componentRef.formatOptions['precision'] = numberDecimalDigits;
            }
            // 小数点符号
            if (numberDecimalSeparator !== null) {
                this.componentRef.decimalSeparator = numberDecimalSeparator;
                this.componentRef.formatOptions['decimalSeparator'] = numberDecimalSeparator;
            }
            // 千分位符号
            if (numberGroupSeparator !== null) {
                this.componentRef.groupSeparator = numberGroupSeparator;
                this.componentRef.formatOptions['groupSeparator'] = numberGroupSeparator;
            }
            this.componentRef.updateValue(this.componentRef.value);
        }
    }
}
FarrisNumberLocalizationDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-number-localization]'
            },] }
];
/** @nocollapse */
FarrisNumberLocalizationDirective.ctorParameters = () => [
    { type: Injector },
    { type: undefined, decorators: [{ type: Inject, args: [UserSettingsToken,] }] },
    { type: NumberSpinnerComponent }
];
