// @farris/kendo-binding/esm2015/lib/directives/localization/farris_date_localization.directive.js
import { Directive, Inject, Injector } from "@angular/core";
import { UserSettingsToken } from "@farris/devkit";
import { FarrisDatepickerComponent } from '@farris/ui-datepicker';
export class FarrisDateLocalizationDirective {
    constructor(injector, userSettings, componentRef) {
        this.injector = injector;
        this.userSettings = userSettings;
        this.componentRef = componentRef;
    }
    ngOnInit() {
        this.localize();
    }
    localize() {
        if (this.userSettings && this.componentRef) {
            const { dateFormat = null, timeFormat = null } = this.userSettings;
            if (dateFormat === null && timeFormat === null) {
                return;
            }
            if (timeFormat && timeFormat && this.componentRef.showTime === true) {
                this.componentRef.dateFormat = `${dateFormat} ${timeFormat}`;
                this.componentRef.dateOpts.dateFormat = `${dateFormat} ${timeFormat}`;
                this.componentRef.updateValue(this.componentRef.value);
            }
            else if (dateFormat) {
                this.componentRef.dateFormat = `${dateFormat}`;
                this.componentRef.dateOpts.dateFormat = `${dateFormat}`;
                this.componentRef.updateValue(this.componentRef.value);
            }
        }
    }
}
FarrisDateLocalizationDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-date-localization]'
            },] }
];
/** @nocollapse */
FarrisDateLocalizationDirective.ctorParameters = () => [
    { type: Injector },
    { type: undefined, decorators: [{ type: Inject, args: [UserSettingsToken,] }] },
    { type: FarrisDatepickerComponent }
];
