// @farris/kendo-binding/esm2015/lib/directives/localization/farris_time_localization.directive.js
import { Directive, Inject, Injector } from "@angular/core";
import { UserSettingsToken } from "@farris/devkit";
import { TimePickerComponent } from '@farris/ui-time-picker';
export class FarrisTimeLocalizationDirective {
    constructor(injector, userSettings, componentRef) {
        this.injector = injector;
        this.userSettings = userSettings;
        this.componentRef = componentRef;
    }
    ngOnInit() {
        this.localize();
    }
    localize() {
        if (this.userSettings && this.componentRef) {
            const { timeFormat = null } = this.userSettings;
            if (timeFormat) {
                this.componentRef.format = timeFormat;
            }
        }
    }
}
FarrisTimeLocalizationDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-time-localization]'
            },] }
];
/** @nocollapse */
FarrisTimeLocalizationDirective.ctorParameters = () => [
    { type: Injector },
    { type: undefined, decorators: [{ type: Inject, args: [UserSettingsToken,] }] },
    { type: TimePickerComponent }
];
