// @farris/mobile-command-services/esm2015/lib/entity-services/entity-traversing.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity-services/entity-traversing.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BindingList } from '@farris/mobile-devkit';
/**
 * 实体遍历服务
 */
class EntityTraversingService {
    /**
     * @param {?} viewModelContext
     */
    constructor(viewModelContext) {
        this.viewModelContext = viewModelContext;
    }
    /**
     * @private
     * @return {?}
     */
    get bindingData() {
        return this.viewModelContext.bindingData;
    }
    /**
     * 获取值
     * @param {?} path
     * @return {?}
     */
    getPropValue(path) {
        return this.bindingData.getValue(path);
    }
    /**
     * 获取实体数据
     * @param {?} path
     * @return {?}
     */
    getEntityData(path) {
        /** @type {?} */
        let bindingObj;
        /** @type {?} */
        const bindingListOrObject = this.bindingData.getValue(path);
        if (bindingListOrObject instanceof BindingList === true) {
            bindingObj = bindingListOrObject.currentItem;
        }
        else {
            bindingObj = bindingListOrObject;
        }
        return bindingObj.toJSON();
    }
    /**
     * 获取实体列表数据
     * @param {?} path
     * @return {?}
     */
    getEntityListData(path) {
        /** @type {?} */
        const bindingList = (/** @type {?} */ (this.bindingData.getValue(path)));
        return bindingList.toJSON();
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    EntityTraversingService.prototype.viewModelContext;
}
export { EntityTraversingService };
