// @farris/mobile-command-services/esm2015/lib/entity-services/entity-aggregation.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity-services/entity-aggregation.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BindingList } from '@farris/mobile-devkit';
import { EntityTraversingService } from './entity-traversing.service';
/**
 * @record
 */
function SplitedPath() { }
if (false) {
    /** @type {?} */
    SplitedPath.prototype.listPath;
    /** @type {?} */
    SplitedPath.prototype.propPath;
}
/**
 * 实体聚合服务
 */
class EntityAggregationService {
    /**
     * @param {?} viewModelContext
     */
    constructor(viewModelContext) {
        this.viewModelContext = viewModelContext;
        this.traversingService = this.viewModelContext.injector.get(EntityTraversingService);
    }
    /**
     * @private
     * @return {?}
     */
    get bindingData() {
        return this.viewModelContext.bindingData;
    }
    /**
     * 获取path对应列表记录的条数
     * @param {?} path
     * @return {?}
     */
    count(path) {
        /** @type {?} */
        const splitedPath = this.splitPath(path);
        /** @type {?} */
        const entityListData = this.traversingService.getEntityListData(splitedPath.listPath);
        return entityListData.length;
    }
    /**
     * 获取path对应字段的和
     * @param {?} path
     * @return {?}
     */
    sum(path) {
        /** @type {?} */
        const splitedPath = this.splitPath(path);
        /** @type {?} */
        const entityListData = this.traversingService.getEntityListData(splitedPath.listPath);
        /** @type {?} */
        const total = entityListData.reduce((/**
         * @param {?} accTotal
         * @param {?} entityData
         * @return {?}
         */
        (accTotal, entityData) => {
            /** @type {?} */
            let propValue = parseFloat(this.getPropValue(entityData, splitedPath.propPath));
            propValue = isNaN(propValue) ? 0 : propValue;
            accTotal += propValue;
            return accTotal;
        }), 0);
        return total;
    }
    /**
     * 获取path对应字段的平均值
     * @param {?} path
     * @return {?}
     */
    avg(path) {
        /** @type {?} */
        const count = this.count(path);
        /** @type {?} */
        const total = this.sum(path);
        return count !== 0 ? total / count : 0;
    }
    /**
     * 获取path对应字段的最大值
     * @param {?} path
     * @return {?}
     */
    max(path) {
        /** @type {?} */
        const splitedPath = this.splitPath(path);
        /** @type {?} */
        const entityListData = this.traversingService.getEntityListData(splitedPath.listPath);
        /** @type {?} */
        const max = entityListData.reduce((/**
         * @param {?} accMax
         * @param {?} entityData
         * @return {?}
         */
        (accMax, entityData) => {
            /** @type {?} */
            const propValue = parseFloat(this.getPropValue(entityData, splitedPath.propPath));
            if (isNaN(propValue) === false && (!accMax || propValue > accMax)) {
                accMax = propValue;
            }
            return accMax;
        }), null);
        return max ? max : 0;
    }
    /**
     * 获取path对应字段的最小值
     * @param {?} path
     * @return {?}
     */
    min(path) {
        /** @type {?} */
        const splitedPath = this.splitPath(path);
        /** @type {?} */
        const entityListData = this.traversingService.getEntityListData(splitedPath.listPath);
        /** @type {?} */
        const min = entityListData.reduce((/**
         * @param {?} accMin
         * @param {?} entityData
         * @return {?}
         */
        (accMin, entityData) => {
            /** @type {?} */
            const propValue = parseFloat(this.getPropValue(entityData, splitedPath.propPath));
            if (isNaN(propValue) === false && (!accMin || propValue < accMin)) {
                accMin = propValue;
            }
            return accMin;
        }), null);
        return min ? min : 0;
    }
    /**
     * 获取属性值
     * @private
     * @param {?} entityData
     * @param {?} path
     * @return {?}
     */
    getPropValue(entityData, path) {
        /** @type {?} */
        let currentData = entityData;
        path.forEach((/**
         * @param {?} pathItem
         * @return {?}
         */
        (pathItem) => {
            // 提前退出的都重置为null，主要针对关联对象未赋值场景
            if (!currentData) {
                currentData = null;
                return;
            }
            currentData = currentData[pathItem];
        }));
        return currentData;
    }
    /**
     * 分隔路径，分隔符列表路径和属性路径
     * @private
     * @param {?} path
     * @return {?} 形如 {parentListPath: [], propertyPath: []}
     */
    splitPath(path) {
        /** @type {?} */
        const listPath = path.concat([]);
        /** @type {?} */
        const propPath = [];
        /** @type {?} */
        let currentNode = this.bindingData.getValue(listPath);
        while (currentNode instanceof BindingList !== true) {
            /** @type {?} */
            const propPathItem = listPath.pop();
            if (!propPathItem) {
                return;
            }
            propPath.unshift(propPathItem);
            currentNode = this.bindingData.getValue(listPath);
        }
        return {
            listPath: listPath,
            propPath: propPath
        };
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    EntityAggregationService.prototype.traversingService;
    /**
     * @type {?}
     * @private
     */
    EntityAggregationService.prototype.viewModelContext;
}
export { EntityAggregationService };
