// @edp-aif/runtime-api/esm2015/lib/aif-manual-creation/service/aif-source-ui.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, ComponentFactoryResolver, NgModuleFactoryLoader, Inject, LOCALE_ID } from "@angular/core";
import { BsModalService } from '@farris/ui-modal';
import { RpcCreationRuleExecutorResult } from '../../entity';
import { AifDataService } from './aif-data.service';
import { FrameContext } from '@farris/devkit';
import { NavigationService } from '@farris/command-services';
import { FormGetCreationRuleEntityExArgs } from '../form-args';
import { PushSelectorComponent } from '../components/push-selector/push-selector.component';
import { AIF_UISTATE_KEY_CREATION_RULE_RESULT } from '../config/aif-form-constants';
import { AbstractAifUiService } from './abstract-aif-ui-service';
export class AifSourceUiService extends AbstractAifUiService {
    /**
     * @param {?} aifDataService
     * @param {?} navigationService
     * @param {?} frameContext
     * @param {?} modalService
     * @param {?} injector
     * @param {?} cfr
     * @param {?} loader
     * @param {?} localeId
     */
    constructor(aifDataService, navigationService, frameContext, modalService, injector, cfr, loader, localeId) {
        super(frameContext, modalService, injector, cfr, loader, localeId);
        this.aifDataService = aifDataService;
        this.navigationService = navigationService;
        this.frameContext = frameContext;
        this.modalService = modalService;
        this.injector = injector;
        this.cfr = cfr;
        this.loader = loader;
        this.localeId = localeId;
        this.aifUIStateCreationRuleResultKey = AIF_UISTATE_KEY_CREATION_RULE_RESULT;
    }
    /**
     * 下推时打开生单规则选择器
     * @param {?} sourceVoId
     * @param {?=} bizFlowchartId
     * @param {?=} modalOptions
     * @param {?=} customSelectorModuleUrl
     * @return {?}
     */
    openCreationRuleSelector4Push(sourceVoId, bizFlowchartId, modalOptions, customSelectorModuleUrl) {
        /** @type {?} */
        const args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        args.bizFlowchartId = bizFlowchartId;
        return this.openCreationRuleSelector(args, PushSelectorComponent, modalOptions, customSelectorModuleUrl);
    }
    /**
     * 特定下推时打开生单规则选择器
     * @param {?} sourceVoId
     * @param {?} targetVoId
     * @param {?=} bizFlowchartId
     * @param {?=} modalOptions
     * @param {?=} customSelectorModuleUrl
     * @return {?}
     */
    openCreationRuleSelector4SpecificPush(sourceVoId, targetVoId, bizFlowchartId, modalOptions, customSelectorModuleUrl) {
        /** @type {?} */
        const args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        args.targetVoId = targetVoId;
        args.bizFlowchartId = bizFlowchartId;
        return this.openCreationRuleSelector(args, PushSelectorComponent, modalOptions, customSelectorModuleUrl);
    }
    /**
     * 打开下游单据制单界面
     * @param {?} formSettings
     * @param {?} rpcCreationRuleExecutorResult
     * @param {?=} extParams
     * @return {?}
     */
    openTargetForm(formSettings, rpcCreationRuleExecutorResult, extParams) {
        if (formSettings == null) {
            throw "参数formSettings不可为空";
        }
        if (rpcCreationRuleExecutorResult == null) {
            throw "参数rpcCreationRuleExecutorResult不可为空";
        }
        if (typeof rpcCreationRuleExecutorResult == "string") {
            if (rpcCreationRuleExecutorResult == "") {
                throw "参数rpcCreationRuleExecutorResult不可为空";
            }
            /** @type {?} */
            const temp = new RpcCreationRuleExecutorResult();
            temp.LoadFromJson(rpcCreationRuleExecutorResult);
            rpcCreationRuleExecutorResult = temp;
        }
        if ((rpcCreationRuleExecutorResult instanceof RpcCreationRuleExecutorResult) == false) {
            /** @type {?} */
            const res = new RpcCreationRuleExecutorResult();
            res.LoadFromJsonObject(rpcCreationRuleExecutorResult);
            rpcCreationRuleExecutorResult = res;
        }
        if (rpcCreationRuleExecutorResult.mappingResults == null || rpcCreationRuleExecutorResult.mappingResults.length == 0) {
            return;
        }
        else if (rpcCreationRuleExecutorResult.mappingResults.length == 1) {
            //只有一条数据则打开卡片
            this.navigationService.openApp(rpcCreationRuleExecutorResult.mappingResults[0].associations[0].tarDataId, formSettings.targetCardAppId, formSettings.targetCardAppEntrance, this.buildRouteParams(rpcCreationRuleExecutorResult, extParams == null ? {} : extParams), false);
        }
        else {
            if (formSettings.targetListAppId != null) {
                // 多条数据且配置批量编辑界面
                throw "暂不支持打开批量编辑界面";
                this.navigationService.openApp(rpcCreationRuleExecutorResult.mappingResults[0].associations[0].tarDataId, formSettings.targetListAppId, formSettings.targetListAppEntrance, this.buildRouteParams(rpcCreationRuleExecutorResult, extParams == null ? {} : extParams), false);
            }
            else {
                //有多条数据但未配置批量编辑界面则打开多个卡片
                for (let mappingResult of rpcCreationRuleExecutorResult.mappingResults) {
                    /** @type {?} */
                    let singleRes = new RpcCreationRuleExecutorResult();
                    singleRes.bizFlowchartId = rpcCreationRuleExecutorResult.bizFlowchartId;
                    singleRes.creationRuleId = rpcCreationRuleExecutorResult.creationRuleId;
                    singleRes.creationRuleInstanceId = rpcCreationRuleExecutorResult.creationRuleInstanceId;
                    singleRes.mappingResults = [mappingResult];
                    this.navigationService.openApp(mappingResult.associations[0].tarDataId, formSettings.targetCardAppId, formSettings.targetCardAppEntrance, this.buildRouteParams(singleRes, extParams == null ? {} : extParams), false);
                }
            }
        }
    }
    /**
     *
     * @private
     * @param {?} rpcCreationRuleExecutorResult
     * @param {?} extParams
     * @return {?}
     */
    buildRouteParams(rpcCreationRuleExecutorResult, extParams) {
        /** @type {?} */
        const params = extParams;
        params[this.aifUIStateCreationRuleResultKey] = rpcCreationRuleExecutorResult;
        return params;
    }
}
AifSourceUiService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AifSourceUiService.ctorParameters = () => [
    { type: AifDataService },
    { type: NavigationService },
    { type: FrameContext },
    { type: BsModalService },
    { type: Injector },
    { type: ComponentFactoryResolver },
    { type: NgModuleFactoryLoader },
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AifSourceUiService.prototype.aifUIStateCreationRuleResultKey;
    /**
     * @type {?}
     * @protected
     */
    AifSourceUiService.prototype.aifDataService;
    /**
     * @type {?}
     * @protected
     */
    AifSourceUiService.prototype.navigationService;
    /**
     * @type {?}
     * @protected
     */
    AifSourceUiService.prototype.frameContext;
    /**
     * @type {?}
     * @protected
     */
    AifSourceUiService.prototype.modalService;
    /**
     * @type {?}
     * @protected
     */
    AifSourceUiService.prototype.injector;
    /**
     * @type {?}
     * @protected
     */
    AifSourceUiService.prototype.cfr;
    /**
     * @type {?}
     * @protected
     */
    AifSourceUiService.prototype.loader;
    /** @type {?} */
    AifSourceUiService.prototype.localeId;
}
