// @farris/oa-command-services/esm5/lib/utils/sync-task-shchdule.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Subject } from "rxjs";
/** @enum {string} */
var ScheduleResultType = {
    Finally: "finally",
    Progress: "progress",
    Error: "error",
};
export { ScheduleResultType };
/**
 * @record
 */
export function ScheduleResult() { }
if (false) {
    /** @type {?} */
    ScheduleResult.prototype.type;
    /** @type {?} */
    ScheduleResult.prototype.recentResult;
    /** @type {?} */
    ScheduleResult.prototype.progressIndex;
    /** @type {?} */
    ScheduleResult.prototype.resultStack;
}
var SyncTaskSchdule = /** @class */ (function () {
    function SyncTaskSchdule() {
        this.actIndex = 0;
        this.resultStack = [];
        this.status = 0; // 0 停止中 1 运行中
        // 0 停止中 1 运行中
        this.taskResult = new Subject();
        this.taskStack = [];
    }
    /**
     * @param {?} fun
     * @param {?=} params
     * @param {?=} context
     * @param {?=} key
     * @return {?}
     */
    SyncTaskSchdule.prototype.propTask = /**
     * @param {?} fun
     * @param {?=} params
     * @param {?=} context
     * @param {?=} key
     * @return {?}
     */
    function (fun, params, context, key) {
        var _this = this;
        if (params === void 0) { params = []; }
        this.taskStack.push({
            funName: key,
            fun: fun,
            params: params,
            context: context
        });
        // 未开始执行 需要手动吊起一次
        if (!this.status) {
            this.actTask().subscribe(function (res) {
                _this.taskResult.next(res);
            });
        }
        return this.taskResult;
    };
    /**
     * @return {?}
     */
    SyncTaskSchdule.prototype.actTask = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.status = 1;
        /** @type {?} */
        var result = new Subject();
        /** @type {?} */
        var actOne = this.taskStack[this.actIndex];
        try {
            actOne.fun.apply(actOne.context || window, actOne.params || []).subscribe(function (res) {
                _this.resultStack.push(res);
                result.next({
                    type: _this.actIndex == _this.taskStack.length - 1 ? ScheduleResultType.Finally : ScheduleResultType.Progress,
                    recentResult: res,
                    progressIndex: _this.actIndex,
                    resultStack: _this.resultStack
                });
                if (_this.actIndex < _this.taskStack.length - 1) {
                    _this.actIndex++;
                    _this.actTask();
                }
                else {
                    _this.status = 0;
                }
            });
        }
        catch (error) {
            this.resultStack.push(error);
            result.next({
                type: ScheduleResultType.Error,
                recentResult: error,
                progressIndex: this.actIndex,
                resultStack: this.resultStack
            });
            if (this.actIndex < this.taskStack.length - 1) {
                this.actIndex++;
                this.actTask();
            }
            else {
                this.status = 0;
            }
        }
        return result;
    };
    return SyncTaskSchdule;
}());
export { SyncTaskSchdule };
if (false) {
    /** @type {?} */
    SyncTaskSchdule.prototype.actIndex;
    /** @type {?} */
    SyncTaskSchdule.prototype.resultStack;
    /** @type {?} */
    SyncTaskSchdule.prototype.status;
    /** @type {?} */
    SyncTaskSchdule.prototype.taskResult;
    /** @type {?} */
    SyncTaskSchdule.prototype.taskStack;
}
