// @gsp-wf/wf-process-editor/esm5/lib/views/performer/views/reporting-line/service/wf-report-def-help.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { map } from 'rxjs/operators';
var WfReportDefHelpService = /** @class */ (function () {
    function WfReportDefHelpService(http) {
        this.http = http;
    }
    /**
     * @param {?} url
     * @param {?=} params
     * @return {?}
     */
    WfReportDefHelpService.prototype.getData = /**
     * @param {?} url
     * @param {?=} params
     * @return {?}
     */
    function (url, params) {
        var e_1, _a;
        /** @type {?} */
        var customData = params.customData;
        /** @type {?} */
        var index = params.pageIndex || 1;
        /** @type {?} */
        var size = params.pageSize || 20;
        /** @type {?} */
        var searchValue = JSON.parse(params.searchValue);
        if (searchValue.searchValue || customData.filter) {
            /** @type {?} */
            var query = '{';
            if (searchValue.searchValue) {
                if (searchValue.searchField === '*') {
                    query += "\"codeOrNameLike\":\"" + searchValue.searchValue + "\",";
                }
                else {
                    query += "\"" + searchValue.searchField + "\":\"" + searchValue.searchValue + "\",";
                }
            }
            if (customData.filter) {
                /** @type {?} */
                var keys = Object.keys(customData.filter);
                if (keys && keys.length > 0) {
                    try {
                        for (var keys_1 = tslib_1.__values(keys), keys_1_1 = keys_1.next(); !keys_1_1.done; keys_1_1 = keys_1.next()) {
                            var key = keys_1_1.value;
                            if (customData.filter[key]) {
                                query += "\"" + key + "\":\"" + customData.filter[key] + "\",";
                            }
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (keys_1_1 && !keys_1_1.done && (_a = keys_1.return)) _a.call(keys_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                }
            }
            if (query.substr(query.length - 1, 1) === ',') {
                query = query.substring(0, query.length - 1);
            }
            query += '}';
            query = encodeURIComponent(query);
            query += "&pageIndex=" + index + "&pageSize=" + size;
            url += '?pageFilter=' + query;
        }
        else {
            url += "?pageIndex=" + index + "&pageSize=" + size;
        }
        return this.http.get(url).pipe(map((/**
         * @param {?} re
         * @return {?}
         */
        function (re) { return ({
            items: re.content,
            total: re.totalElements,
            pageInfo: {
                pageIndex: re.pageable.pageNumber + 1,
                pageSize: re.pageable.pageSize,
                pageList: [20, 40, 60, 80, 100],
                enablePager: true
            }
        }); })));
    };
    WfReportDefHelpService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    WfReportDefHelpService.ctorParameters = function () { return [
        { type: HttpService }
    ]; };
    return WfReportDefHelpService;
}());
export { WfReportDefHelpService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    WfReportDefHelpService.prototype.http;
}
