// @gsp-wf/wf-process-editor/esm5/lib/views/performer/views/reporting-line/service/wf-report-def-help.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild } from '@angular/core';
import { of } from 'rxjs';
import { LookupGridComponent, ServerSideToken } from '@farris/ui-lookup';
import { WfReportDefHelpService } from './wf-report-def-help.service';
import { WfLocalePipe } from '../../../../../i18n/pipe/wf-process-editor-locale.pipe';
var WfReportDefHelpComponent = /** @class */ (function () {
    function WfReportDefHelpComponent(localePipe) {
        var _this = this;
        this.localePipe = localePipe;
        /**
         * 宽度
         */
        this.width = 550;
        /**
         * 只读
         */
        this.disabled = false;
        /**
         * 是否单选
         */
        this.singleSelect = true;
        /**
         * 是否分页，默认true
         */
        this.pagination = true;
        /**
         * 首次加载第几页，默认1
         */
        this.pageIndex = 1;
        /**
         * 每页加载数，默认20
         */
        this.pageSize = 20;
        /**
         * 过滤条件
         */
        this.filter = {};
        // @Input() beforeClose: (selectData: any) => Observable<PickedResult>;
        /**
         * 帮助选择后事件
         */
        this.afterConfirm = new EventEmitter();
        this.dialogClosed = new EventEmitter();
        this.clear = new EventEmitter();
        this.columns = [];
        /**
         * 帮助前事件,强行传递过滤条件
         */
        this.beforeOpen = (/**
         * @param {?} arg
         * @return {?}
         */
        function (arg) {
            // 把过滤数据和当前的数据传递到service层
            return of({
                showDialog: true, data: {
                    filter: _this.filter
                }
            });
        });
        this.dictPicked = (/**
         * @param {?} arg
         * @return {?}
         */
        function (arg) {
            _this.afterConfirm.next(arg);
            return of({ closeDialog: true });
        });
    }
    /**
     * @return {?}
     */
    WfReportDefHelpComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.columns = [
            { title: this.localePipe.transform('component.reportingLine.code'), field: 'code', width: '80', searchField: 'code', },
            { title: this.localePipe.transform('component.reportingLine.name'), field: 'name', width: '100', searchField: 'name', }
        ];
        // this.dataUri = this.uiState.dataUri;
        this.dataUri = '/api/runtime/wf/v1.0/reportingLineDefinitions/query';
        // 必须要有，否则初始选中不起作用
        this.lookUp.mapFields = { id: 'id', name: 'name' };
    };
    /**
     * @param {?} arg
     * @return {?}
     */
    WfReportDefHelpComponent.prototype.afterClear = /**
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        this.clear.next(arg);
    };
    /** 手动显示帮助 */
    /**
     * 手动显示帮助
     * @return {?}
     */
    WfReportDefHelpComponent.prototype.showHelp = /**
     * 手动显示帮助
     * @return {?}
     */
    function () {
        this.lookUp.showDialog();
    };
    WfReportDefHelpComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-wf-report-def-help',
                    template: "\n    <farris-lookup-grid #lookup\n      [disabled]=\"disabled\"\n      [editable]=\"false\"\n      [bindingData]=\"bindingData\"\n      [(ngModel)]=\"displayTxt\"\n      [columns]=\"columns\"\n      [singleSelect]=\"singleSelect\"\n      title=\"title\"\n      [displayType]=\"'List'\"\n      [uri]=\"dataUri\"\n      [idField]=\"'id'\"\n      [pageIndex]=\"pageIndex\"\n      [pageSize]=\"pageSize\"\n      [pagination]=\"pagination\"\n      [valueField]=\"'id'\"\n      [textField]=\"'name'\"\n      (clear)=\"afterClear($event)\"\n      [dialogWidth]=\"width\"\n      [dictPicking]=\"beforeOpen\"\n      [dictPicked]=\"dictPicked\">\n    </farris-lookup-grid>\n  ",
                    providers: [{ provide: ServerSideToken, useClass: WfReportDefHelpService }]
                }] }
    ];
    /** @nocollapse */
    WfReportDefHelpComponent.ctorParameters = function () { return [
        { type: WfLocalePipe }
    ]; };
    WfReportDefHelpComponent.propDecorators = {
        title: [{ type: Input }],
        width: [{ type: Input }],
        sourceType: [{ type: Input }],
        disabled: [{ type: Input }],
        singleSelect: [{ type: Input }],
        pagination: [{ type: Input }],
        pageIndex: [{ type: Input }],
        pageSize: [{ type: Input }],
        displayTxt: [{ type: Input }],
        bindingData: [{ type: Input }],
        filter: [{ type: Input }],
        afterConfirm: [{ type: Output }],
        dialogClosed: [{ type: Output }],
        clear: [{ type: Output }],
        lookUp: [{ type: ViewChild, args: ['lookup',] }]
    };
    return WfReportDefHelpComponent;
}());
export { WfReportDefHelpComponent };
if (false) {
    /**
     * 标题
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.title;
    /**
     * 宽度
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.width;
    /**
     * 取数源
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.sourceType;
    /**
     * 只读
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.disabled;
    /**
     * 是否单选
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.singleSelect;
    /**
     * 是否分页，默认true
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.pagination;
    /**
     * 首次加载第几页，默认1
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.pageIndex;
    /**
     * 每页加载数，默认20
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.pageSize;
    /**
     * 绑定到帮助上的数据，初始显示值用
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.displayTxt;
    /**
     * 存储帮助选择的对象
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.bindingData;
    /**
     * 过滤条件
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.filter;
    /**
     * 帮助选择后事件
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.afterConfirm;
    /** @type {?} */
    WfReportDefHelpComponent.prototype.dialogClosed;
    /** @type {?} */
    WfReportDefHelpComponent.prototype.clear;
    /** @type {?} */
    WfReportDefHelpComponent.prototype.dataUri;
    /** @type {?} */
    WfReportDefHelpComponent.prototype.columns;
    /** @type {?} */
    WfReportDefHelpComponent.prototype.lookUp;
    /**
     * 帮助前事件,强行传递过滤条件
     * @type {?}
     */
    WfReportDefHelpComponent.prototype.beforeOpen;
    /** @type {?} */
    WfReportDefHelpComponent.prototype.dictPicked;
    /**
     * @type {?}
     * @private
     */
    WfReportDefHelpComponent.prototype.localePipe;
}
