// @farris/ui-dropdown/esm5/lib/f-dropdown-toggle.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostBinding, HostListener, ElementRef, Input, } from "@angular/core";
import { FDropdownDirective } from "./f-dropdown.directive";
var FDropdownToggleDirective = /** @class */ (function () {
    function FDropdownToggleDirective(dropdown, elementRef) {
        var _this = this;
        this.dropdown = dropdown;
        this.elementRef = elementRef;
        this._disabled = false;
        this.haspopup = true;
        this.nativeEle = null;
        dropdown.toggleElement = elementRef.nativeElement;
        this.nativeEle = elementRef.nativeElement;
        this.dropdown.getOpenState().subscribe((/**
         * @param {?} state
         * @return {?}
         */
        function (state) {
            if (_this.dropdown.isSubDP && _this.dropdown.isOpen) {
                if (_this.nativeEle.className.indexOf("active") < 0) {
                    _this.nativeEle.className += " active";
                }
            }
            else {
                if (_this.nativeEle.className.indexOf("active") > -1) {
                    _this.nativeEle.className = _this.nativeEle.className.replace(" active", " ");
                }
            }
        }));
    }
    Object.defineProperty(FDropdownToggleDirective.prototype, "dpDisabled", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FDropdownToggleDirective.prototype, "disableCls", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} ev
     * @return {?}
     */
    FDropdownToggleDirective.prototype.onclick = /**
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        if (this._disabled) {
            return;
        }
        if (this.dropdown.isSubDP) {
            ev.stopImmediatePropagation();
        }
        this.dropdown.toggle();
    };
    FDropdownToggleDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[fDropdownToggle]",
                },] }
    ];
    /** @nocollapse */
    FDropdownToggleDirective.ctorParameters = function () { return [
        { type: FDropdownDirective },
        { type: ElementRef }
    ]; };
    FDropdownToggleDirective.propDecorators = {
        dpDisabled: [{ type: Input }],
        disableCls: [{ type: HostBinding, args: ["class.disabled",] }],
        haspopup: [{ type: HostBinding, args: ["attr.aria-haspopup",] }],
        onclick: [{ type: HostListener, args: ["click", ["$event"],] }]
    };
    return FDropdownToggleDirective;
}());
export { FDropdownToggleDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FDropdownToggleDirective.prototype._disabled;
    /** @type {?} */
    FDropdownToggleDirective.prototype.haspopup;
    /**
     * @type {?}
     * @private
     */
    FDropdownToggleDirective.prototype.nativeEle;
    /**
     * @type {?}
     * @private
     */
    FDropdownToggleDirective.prototype.dropdown;
    /**
     * @type {?}
     * @private
     */
    FDropdownToggleDirective.prototype.elementRef;
}
