// @gsp-wf/wf-process-model/esm5/lib/model/mex/CDTSchemaType.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, CustomType } from "@edp-pmf/bpmn-model";
var CDTSchemaType = /** @class */ (function (_super) {
    tslib_1.__extends(CDTSchemaType, _super);
    function CDTSchemaType(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = CDTSchemaType.Clr_Type_ID;
        return _this;
    }
    /**
     * @return {?}
     */
    CDTSchemaType.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new CDTSchemaType(this.Model);
    };
    /**
     * @return {?}
     */
    CDTSchemaType.prototype.ResetBeforeLoadFrom = /**
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.MetadataId = null;
    };
    /**
     * @return {?}
     */
    CDTSchemaType.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "SchemaId", this.SchemaId);
        BpmnModelHelper.SetAttributeValue(obj, "MetadataId", this.MetadataId);
        BpmnModelHelper.SetAttributeValue(obj, "ComponentId", this.ComponentId);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CDTSchemaType.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.SchemaId = obj.SchemaId;
        this.MetadataId = obj.MetadataId;
        this.ComponentId = obj.ComponentId;
    };
    CDTSchemaType.Clr_Type_ID = "wf.CDTSchemaType";
    return CDTSchemaType;
}(CustomType));
export { CDTSchemaType };
if (false) {
    /** @type {?} */
    CDTSchemaType.Clr_Type_ID;
    /** @type {?} */
    CDTSchemaType.prototype.ClrTypeID;
    /** @type {?} */
    CDTSchemaType.prototype.SchemaId;
    /** @type {?} */
    CDTSchemaType.prototype.MetadataId;
    /** @type {?} */
    CDTSchemaType.prototype.ComponentId;
}
BpmnModelHelper.RegisterElementType(new CDTSchemaType(new BpmnModel()));
