// @gsp-wf/wf-process-model/esm5/lib/model/mex/MexExpressionImpl.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, MexExpression } from "@edp-pmf/bpmn-model";
var MexExpressionImpl = /** @class */ (function (_super) {
    tslib_1.__extends(MexExpressionImpl, _super);
    function MexExpressionImpl(model, value) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = MexExpressionImpl.Clr_Type_ID;
        _this.value = value;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    MexExpressionImpl.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.value = '';
    };
    /**
     * @return {?}
     */
    MexExpressionImpl.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, 'Value', this.value);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    MexExpressionImpl.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.value = obj.Value;
    };
    /**
     * @return {?}
     */
    MexExpressionImpl.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new MexExpressionImpl(this.Model);
    };
    /**
     * @return {?}
     */
    MexExpressionImpl.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var mexExpressionImpl = new MexExpressionImpl(this.Model);
        mexExpressionImpl.value = this.value;
        return mexExpressionImpl;
    };
    MexExpressionImpl.Clr_Type_ID = "wf.MexExpressionImpl";
    return MexExpressionImpl;
}(MexExpression));
export { MexExpressionImpl };
if (false) {
    /** @type {?} */
    MexExpressionImpl.Clr_Type_ID;
    /** @type {?} */
    MexExpressionImpl.prototype.ClrTypeID;
    /** @type {?} */
    MexExpressionImpl.prototype.value;
}
BpmnModelHelper.RegisterElementType(new MexExpressionImpl(new BpmnModel()));
