// @gsp-wf/wf-process-model/esm5/lib/model/mex/TitleExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { MexExpressionImpl } from "./MexExpressionImpl";
var TitleExpression = /** @class */ (function (_super) {
    tslib_1.__extends(TitleExpression, _super);
    function TitleExpression(model, value, valueLanguage) {
        var _this = _super.call(this, model, value) || this;
        _this.ClrTypeID = TitleExpression.Clr_Type_ID;
        _this.valueLanguage = {};
        _this.valueLanguage = valueLanguage;
        return _this;
    }
    /**
     * @return {?}
     */
    TitleExpression.prototype.ResetBeforeLoadFrom = /**
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.valueLanguage = {};
    };
    /**
     * @return {?}
     */
    TitleExpression.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, 'ValueLanguage', this.valueLanguage);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    TitleExpression.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.valueLanguage = obj.ValueLanguage;
    };
    /**
     * @return {?}
     */
    TitleExpression.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new TitleExpression(this.Model);
    };
    /**
     * @return {?}
     */
    TitleExpression.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var titleExpression = new TitleExpression(this.Model);
        titleExpression.value = this.value;
        titleExpression.valueLanguage = this.valueLanguage;
        return titleExpression;
    };
    TitleExpression.Clr_Type_ID = 'wf.TitleExpression';
    return TitleExpression;
}(MexExpressionImpl));
export { TitleExpression };
if (false) {
    /** @type {?} */
    TitleExpression.Clr_Type_ID;
    /** @type {?} */
    TitleExpression.prototype.ClrTypeID;
    /** @type {?} */
    TitleExpression.prototype.valueLanguage;
}
BpmnModelHelper.RegisterElementType(new TitleExpression(new BpmnModel()));
