// @gsp-wf/rtdevkit/esm2015/lib/runtime/components/task-transfer/task-transfer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { TaskService } from '../../services/task.service';
import { UserService } from '../../services/user.service';
import { UserHelperComponent } from '../user-helper/user-helper.component';
import { NotifyService } from '@farris/ui-notify';
import { I18nService } from '../../services/i18n.service';
export class TaskTransferComponent extends UserHelperComponent {
    /**
     * @param {?} taskSvc
     * @param {?} userSvc
     * @param {?} notify
     * @param {?} i18nService
     */
    constructor(taskSvc, userSvc, notify, i18nService) {
        super(i18nService, userSvc);
        this.taskSvc = taskSvc;
        this.userSvc = userSvc;
        this.notify = notify;
        this.i18nService = i18nService;
        this.completed = new EventEmitter();
        this.singleSelect = true;
        this.paginationOptions.id = 'WF-TaskTransfer-DataTable-Pagination';
    }
    /**
     * @param {?} context
     * @return {?}
     */
    set transfer_context(context) {
        this.transferContext = Object.assign({}, context, { assignee: '' });
        this.selectedUsers = [];
        this.showModal(!!this.transferContext.workItemId);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    handleOk() {
        this.selectedUsers = this.userTable.selections;
        // 需要校验选中唯一用户
        if (!this.selectedUsers) {
            this.notify.info(this.i18nService.getValue('static.transfer.selectHandover'));
            return;
        }
        this.transferContext.assignee = this.selectedUsers['ID'];
        this.taskSvc.transfer(this.transferContext).subscribe((/**
         * @return {?}
         */
        () => {
            this.notify.success(this.i18nService.getValue('static.transfer.successTransfer'));
            this.completed.emit(true);
            this.showModal(false);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            if (error && error.error && error.error.Message) {
                this.notify.error(error.error.Message);
            }
        }));
    }
}
TaskTransferComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf-task-transfer',
                template: "<farris-dialog #allUserDialog [buttons]=\"buttonRef\" [width]=\"850\" [height]=\"520\" [title]=\"'\u7CFB\u7EDF\u7528\u6237'\">\r\n  <div class=\"d-flex flex-column my-data-grid h-100\" style=\"padding: 0 14px;\">\r\n    <div class=\"flex-auto\" style=\"position:relative;\">\r\n      <farris-datatable #allUserTable [columns]=\"columns\" [data]=\"userListData | paginate: paginationOptions\"\r\n        [showFilterBar]='true' [singleSelect]=\"singleSelect\" [idField]=\"'ID'\" [remote]=\"'server'\" (search)=\"search($event)\"\r\n        [pagination]='false'></farris-datatable>\r\n    </div>\r\n    <div style=\"position:relative;\">\r\n      <pagination-controls [id]=\"paginationOptions.id\" [previousLabel]=\"' '\" [nextLabel]=\"' '\"\r\n        (pageChange)=\"onPageChange($event)\" (pageSizeChange)=\"onPageSizeChange($event)\">\r\n      </pagination-controls>\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n<ng-template #buttonRef>\r\n  <button class=\"btn btn-sm btn-primary px-3\" (click)=\"handleOk()\">\u786E\u5B9A</button>\r\n  <button class=\"btn btn-sm btn-default px-3\" (click)=\"handleCancel()\">\u53D6\u6D88</button>\r\n</ng-template>",
                styles: [".my-data-grid .filter-row .search-fields{padding-left:0!important}.my-data-grid .filter-row .search-text{padding-right:0!important}"]
            }] }
];
/** @nocollapse */
TaskTransferComponent.ctorParameters = () => [
    { type: TaskService },
    { type: UserService },
    { type: NotifyService },
    { type: I18nService }
];
TaskTransferComponent.propDecorators = {
    transfer_context: [{ type: Input }],
    completed: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    TaskTransferComponent.prototype.completed;
    /** @type {?} */
    TaskTransferComponent.prototype.transferContext;
    /**
     * @type {?}
     * @private
     */
    TaskTransferComponent.prototype.taskSvc;
    /**
     * @type {?}
     * @protected
     */
    TaskTransferComponent.prototype.userSvc;
    /**
     * @type {?}
     * @protected
     */
    TaskTransferComponent.prototype.notify;
    /**
     * @type {?}
     * @protected
     */
    TaskTransferComponent.prototype.i18nService;
}
