// @gsp-cmp/ubcc-param-vo/esm2015/lib/vo/serializer/constraint-serializer/integer-constraint-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/constraint-serializer/integer-constraint-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { UbccIntegerConstraint } from '../../extension/constraint/ubcc-integer-constraint';
export class UbccIntegerConstraintSerializer {
    /**
     * @param {?} constraint
     * @param {?} ctxt
     * @return {?}
     */
    serialize(constraint, ctxt) {
        /** @type {?} */
        let jobj = new Object();
        if (!constraint) {
            return jobj;
        }
        /** @type {?} */
        let integerConstraint = (/** @type {?} */ (constraint));
        jobj['kind'] = integerConstraint.kind;
        jobj['max'] = integerConstraint.max;
        jobj['min'] = integerConstraint.min;
        jobj['unit'] = integerConstraint.unit;
        return jobj;
    }
    /**
     * @param {?} jobj
     * @param {?} ctxt
     * @return {?}
     */
    deserialize(jobj, ctxt) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        let integerConstraint = new UbccIntegerConstraint();
        integerConstraint.max = (/** @type {?} */ (jobj['max']));
        integerConstraint.min = (/** @type {?} */ (jobj['min']));
        integerConstraint.unit = jobj['unit'];
        return integerConstraint;
    }
}
