// @farris/ui-lookup/esm5/lib/advanced-layout/lookup-advanced-layout.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Component, ElementRef, Injector, Input, NgZone, Output, Renderer2, ViewChild, ViewEncapsulation } from '@angular/core';
import { DialogComponent } from '@farris/ui-dialog';
var LookupAdvancedLayoutComponent = /** @class */ (function () {
    function LookupAdvancedLayoutComponent(elRef, injector, zone, render2) {
        this.elRef = elRef;
        this.injector = injector;
        this.zone = zone;
        this.render2 = render2;
        this.title = '帮助选择器';
        this.activeTab = 'datalist';
        this.dataTab = '数据列表';
        this.favTab = '收藏夹';
        this.tabChanged = new EventEmitter();
        this.close = new EventEmitter();
        this.dialogRef = this.injector.get(DialogComponent, null);
    }
    /**
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.activeTab && !changes.activeTab.isFirstChange()) {
            this.setCurrentTab();
        }
    };
    /**
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.dialogRef) {
            this.dialogRef.dragHandle = this.headerRel.nativeElement;
        }
        this.setCurrentTab();
    };
    /**
     * @private
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.setCurrentTab = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.activetabRef) {
            this.render2.setStyle(this.activetabRef.nativeElement, 'transform', "translate3d(" + (this.activeTab == 'datalist' ? 0 : 84) + "px, 0px, 0px)");
        }
    };
    /**
     * @param {?} $event
     * @param {?} tabstr
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.changeTabs = /**
     * @param {?} $event
     * @param {?} tabstr
     * @return {?}
     */
    function ($event, tabstr) {
        this.activeTab = tabstr;
        this.setCurrentTab();
        this.tabChanged.emit({ tab: tabstr, $event: $event });
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.closeDialog = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.dialogRef) {
            $event.stopPropagation();
            this.dialogRef.close();
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.changeDialogSize = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.dialogRef) {
            $event.stopPropagation();
            this.dialogRef.changeDialogSize();
        }
    };
    LookupAdvancedLayoutComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lookup-advanced-layout',
                    template: "<div class=\"d-flex justify-content-between header\" #header [style.cursor]=\"dialogRef.isMax ? 'default': 'move'\">\r\n    <div class=\"title\" style=\"max-width: calc(50% - 100px);text-overflow: ellipsis;overflow: hidden;z-index: 3;cursor: default;\" title=\"{{title}}\">{{title}}</div>\r\n\r\n    <div style=\"width: 100%;\" class=\"tabs\">\r\n        <div class=\"d-flex justify-content-center\">\r\n            <div class=\"d-flex tabs-labels\" *ngIf=\"enableFav\"  (mousedown)=\"$event.stopPropagation()\">\r\n                <div class=\"active-tab\" #activetab></div>\r\n                <div class=\"tab-title d-flex\" [class.active]=\"activeTab === 'datalist'\" (click)=\"changeTabs($event, 'datalist')\" >\r\n                    {{dataTab }}\r\n                </div>\r\n                <div class=\"tab-title d-flex\" [class.active]=\"activeTab === 'favorite'\" (click)=\"changeTabs($event, 'favorite')\">\r\n                    {{favTab}}\r\n                </div>\r\n            </div>\r\n        </div>\r\n    </div>\r\n\r\n    <div class=\"close d-flex flex-row h-100\">\r\n        <span class=\"f-icon modal_maximize\" [class.modalrevert]=\"dialogRef.isMax\" (click)=\"changeDialogSize($event)\"></span>\r\n        <span class=\"f-icon modal_close\" (click)=\"closeDialog($event)\"></span>\r\n    </div>\r\n</div>",
                    // styleUrls: ['./advanced-layout.scss'],
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    LookupAdvancedLayoutComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Injector },
        { type: NgZone },
        { type: Renderer2 }
    ]; };
    LookupAdvancedLayoutComponent.propDecorators = {
        title: [{ type: Input }],
        activeTab: [{ type: Input }],
        enableFav: [{ type: Input }],
        dataTab: [{ type: Input }],
        favTab: [{ type: Input }],
        tabChanged: [{ type: Output }],
        close: [{ type: Output }],
        headerRel: [{ type: ViewChild, args: ['header',] }],
        activetabRef: [{ type: ViewChild, args: ['activetab',] }]
    };
    return LookupAdvancedLayoutComponent;
}());
export { LookupAdvancedLayoutComponent };
if (false) {
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.title;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.activeTab;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.enableFav;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.dataTab;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.favTab;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.tabChanged;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.close;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.headerRel;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.activetabRef;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.dialogRef;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.elRef;
    /**
     * @type {?}
     * @private
     */
    LookupAdvancedLayoutComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    LookupAdvancedLayoutComponent.prototype.zone;
    /**
     * @type {?}
     * @private
     */
    LookupAdvancedLayoutComponent.prototype.render2;
}
