// @gsp-svc/expression/esm5/lib/expression-smart-help-grid/expression-smart-help-grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { ExpressionEditorLocalePipe } from '../expression-editor-locale.pipe';
var ExpressionSmartHelpGridComponent = /** @class */ (function () {
    function ExpressionSmartHelpGridComponent() {
        this.smartHelp = [];
        this.selectSchemaRow = new EventEmitter();
        this.dbclickInsert = new EventEmitter();
        this.items = [];
        this.columns = [];
        //初始化i18n标识
        this.localeId = localStorage.getItem('languageCode');
        this.localePipe = new ExpressionEditorLocalePipe(this.localeId);
    }
    /**
     * @return {?}
     */
    ExpressionSmartHelpGridComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.columns = [
            { field: 'code', width: 60, title: this.localePipe.transform("HelpCode"), filter: true },
            { field: 'name', width: 60, title: this.localePipe.transform("HelpName"), filter: true },
            { field: 'description', width: 100, title: this.localePipe.transform("HelpDescription"), filter: true },
        ];
        this.items = this.smartHelp;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    ExpressionSmartHelpGridComponent.prototype.ondblclickRow = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var entities = this.smartHelp.find((/**
         * @param {?} temp
         * @return {?}
         */
        function (temp) { return temp.code == $event.rowData.code; }));
        // entities.serviceInstance.showWizard().subscribe((result)=>{
        //   this.dbclickInsert.emit(result);
        // });
        this.dbclickInsert.emit(entities.serviceInstance.showWizard());
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    ExpressionSmartHelpGridComponent.prototype.selectRowChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.selectSchemaRow.emit($event.data);
    };
    ExpressionSmartHelpGridComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-expression-smart-help-grid',
                    template: "<farris-datagrid\r\n[idField]=\"'code'\"\r\n[height]=\"268\"\r\n[width]=\"650\"\r\n[columns]=\"columns\" \r\n[data]=\"items\"\r\n[fit]=\"true\"\r\n[fitColumns]=\"true\"\r\n[striped]=\"false\"\r\n[showHeader]=\"true\"\r\n[pagination]=\"false\"\r\n(dblclickRow)=\"ondblclickRow($event)\"\r\n(selectChanged)=\"selectRowChange($event)\"\r\n>\r\n</farris-datagrid>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ExpressionSmartHelpGridComponent.ctorParameters = function () { return []; };
    ExpressionSmartHelpGridComponent.propDecorators = {
        smartHelp: [{ type: Input }],
        selectSchemaRow: [{ type: Output }],
        dbclickInsert: [{ type: Output }]
    };
    return ExpressionSmartHelpGridComponent;
}());
export { ExpressionSmartHelpGridComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ExpressionSmartHelpGridComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    ExpressionSmartHelpGridComponent.prototype.localePipe;
    /** @type {?} */
    ExpressionSmartHelpGridComponent.prototype.smartHelp;
    /** @type {?} */
    ExpressionSmartHelpGridComponent.prototype.selectSchemaRow;
    /** @type {?} */
    ExpressionSmartHelpGridComponent.prototype.dbclickInsert;
    /** @type {?} */
    ExpressionSmartHelpGridComponent.prototype.items;
    /** @type {?} */
    ExpressionSmartHelpGridComponent.prototype.columns;
}
