// @farris/command-services/esm2015/lib/manifest/form_expression_manifest_service.js
import { HttpClient } from "@angular/common/http";
import { Inject, Injectable, Injector } from "@angular/core";
import { of } from "rxjs";
import { share, switchMap } from "rxjs/operators";
import { FORM_PATH_TOKEN, FrameContext, FORM_MANIFEST_SERVICE_TOKEN } from "@farris/devkit";
import { BasePathService } from "@farris/rtf";
export class FormExpressionManifestService {
    constructor(injector, frameContext, httpClient, modulePath, formManifestService) {
        this.injector = injector;
        this.frameContext = frameContext;
        this.httpClient = httpClient;
        this.modulePath = modulePath;
        this.formManifestService = formManifestService;
    }
    load() {
        if (this.modulePath.endsWith('\/')) {
            this.modulePath = this.modulePath.substring(0, this.modulePath.length - 1);
        }
        return this.formManifestService.load().pipe(switchMap((formManifest) => {
            const expressions = formManifest.expressions;
            const expressionManifest = expressions.find((expressionManifest) => expressionManifest.ns === this.frameContext.namespace);
            if (expressionManifest) {
                const expressionPath = BasePathService.convertPath(`${this.modulePath}/expressions/${expressionManifest.path}?version=${new Date().valueOf().toString()}`);
                const key = BasePathService.convertPath(`${this.modulePath}/expressions/${expressionManifest.path}`);
                const request$ = FormExpressionManifestService.mainfests.get(key);
                if (request$) {
                    return request$;
                }
                else {
                    const req$ = this.httpClient.get(expressionPath, { responseType: 'json' }).pipe(share());
                    FormExpressionManifestService.mainfests.set(key, req$);
                    return req$;
                }
            }
            else {
                return of({});
            }
        }));
    }
}
FormExpressionManifestService.mainfests = new Map();
FormExpressionManifestService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FormExpressionManifestService.ctorParameters = () => [
    { type: Injector },
    { type: FrameContext },
    { type: HttpClient },
    { type: undefined, decorators: [{ type: Inject, args: [FORM_PATH_TOKEN,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [FORM_MANIFEST_SERVICE_TOKEN,] }] }
];
