// @gsp-lcm/metadatart-selector/esm2015/metadata/metadatart-select/metadatart.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadatart-select/metadatart.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { GSPMetadataRTService } from '../metadata.rtservice';
export class MetadataRTService {
    /**
     * @param {?} gspMetadataService
     */
    constructor(gspMetadataService) {
        this.gspMetadataService = gspMetadataService;
        this.metadatas = [];
    }
    /**
     * @param {?} metadata
     * @return {?}
     */
    addMetadataRT(metadata) {
        this.metadatas.push(metadata);
    }
    /**
     * @return {?}
     */
    clearMetadataRT() {
        this.metadatas = [];
    }
    // 根据类型获取Cloud Server下所有的元数据
    /**
     * @param {?} typeList
     * @return {?}
     */
    GetMetadataRTList(typeList) {
        return this.gspMetadataService.GetMetadataRTList(typeList);
    }
    /**
     * @param {?} typeList
     * @return {?}
     */
    GetMetadataRTListForHelp(typeList) {
        return this.gspMetadataService.GetMetadataRTListForHelp(typeList);
    }
    /**
     * @param {?} bizTypeId
     * @param {?} metadataTypes
     * @return {?}
     */
    getMetadataListByBizTypeId(bizTypeId, metadataTypes) {
        return this.gspMetadataService.getMetadataListByBizTypeId(bizTypeId, metadataTypes);
    }
    /**
     * @param {?} boList
     * @param {?} bizTypeId
     * @param {?} metadataTypes
     * @return {?}
     */
    getMetadataListByBizTypeIdForHelp(boList, bizTypeId, metadataTypes) {
        if (boList === undefined) {
            boList = '';
        }
        if (bizTypeId === undefined) {
            bizTypeId = '';
        }
        return this.gspMetadataService.getMetadataListByBizTypeIdForHelp(boList, bizTypeId, metadataTypes);
    }
    // 获根据Id取指定元数据
    /**
     * @param {?} metadataID
     * @return {?}
     */
    GetMetadataRT(metadataID) {
        return this.gspMetadataService.GetMetadataRT(metadataID);
    }
}
MetadataRTService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
MetadataRTService.ctorParameters = () => [
    { type: GSPMetadataRTService }
];
if (false) {
    /** @type {?} */
    MetadataRTService.prototype.metadatas;
    /**
     * @type {?}
     * @private
     */
    MetadataRTService.prototype.gspMetadataService;
}
