// @farris/mobile-command-services/esm5/lib/ccwork-service/ccwork.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ccwork-service/ccwork.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
var CcWorkService = /** @class */ (function () {
    // public location$ = new Subject();
    function CcWorkService() {
    }
    /**
     * 云上协同关闭页面
     */
    /**
     * 云上协同关闭页面
     * @return {?}
     */
    CcWorkService.prototype.closeCcWorkWebView = /**
     * 云上协同关闭页面
     * @return {?}
     */
    function () {
        window['iCityBridge'].closeWebView();
    };
    /**
     * 云上协同隐藏标题栏
     */
    /**
     * 云上协同隐藏标题栏
     * @return {?}
     */
    CcWorkService.prototype.hideCcWorkNaviBar = /**
     * 云上协同隐藏标题栏
     * @return {?}
     */
    function () {
        window['iCityBridge'].hideTitleBar();
    };
    /**
     *  云上协同更改标题栏
     */
    /**
     *  云上协同更改标题栏
     * @param {?} title
     * @return {?}
     */
    CcWorkService.prototype.setTitle = /**
     *  云上协同更改标题栏
     * @param {?} title
     * @return {?}
     */
    function (title) {
        window['iCityBridge'].setTitle(title);
    };
    /**
     * 获取当前位置
     * @param callBack(data) 回调函数
     */
    /**
     * 获取当前位置
     * @param {?} callBack
     * @return {?}
     */
    CcWorkService.prototype.getLocation = /**
     * 获取当前位置
     * @param {?} callBack
     * @return {?}
     */
    function (callBack) {
        if (window['iCityBridge']) {
            window['iCityBridge'].ccworkGetLocation(callBack);
        }
    };
    /**
     * 获取二维码
     * @param callBack(data) 回调函数
     * @param params 配置项
     */
    /**
     * 获取二维码
     * @param {?} callBack
     * @param {?=} params 配置项
     * @return {?}
     */
    CcWorkService.prototype.scanCode = /**
     * 获取二维码
     * @param {?} callBack
     * @param {?=} params 配置项
     * @return {?}
     */
    function (callBack, params) {
        if (params === void 0) { params = {}; }
        if (window['iCityBridge']) {
            /** @type {?} */
            var paramsObj = tslib_1.__assign({ filters: [
                    'QR', 'EAN13', 'EAN8'
                ], path: '' }, params);
            window['iCityBridge'].ccworkScanCode(paramsObj, callBack);
        }
    };
    /**
     * 云上协同调用拨号
     * @param phoneNumber 手机号
     * @param callBack(data) 回调函数
     */
    /**
     * 云上协同调用拨号
     * @param {?} phoneNumber 手机号
     * @param {?} callback
     * @return {?}
     */
    CcWorkService.prototype.callPhone = /**
     * 云上协同调用拨号
     * @param {?} phoneNumber 手机号
     * @param {?} callback
     * @return {?}
     */
    function (phoneNumber, callback) {
        if (window['iCityBridge']) {
            /** @type {?} */
            var params = {
                number: phoneNumber
            };
            window['iCityBridge'].ccworkCallPhone(params, callback);
        }
    };
    /**
     * 从相机和相册中获取照片
     * @param params
     * @param callback
     */
    /**
     * 从相机和相册中获取照片
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    CcWorkService.prototype.getPicsFromCameraAndAlbum = /**
     * 从相机和相册中获取照片
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        if (window['iCityBridge']) {
            window['iCityBridge'].ccworkGetPicsFromCameraAndAlbum(params, callback);
        }
    };
    return CcWorkService;
}());
export { CcWorkService };
