// @farris/mobile-command-services/esm5/lib/work-flow-service/service/task.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/work-flow-service/service/task.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var TaskService = /** @class */ (function () {
    function TaskService(httpSvc) {
        this.httpSvc = httpSvc;
    }
    /**
     * @param {?} error
     * @param {?=} actionCode
     * @return {?}
     */
    TaskService.prototype.error = /**
     * @param {?} error
     * @param {?=} actionCode
     * @return {?}
     */
    function (error, actionCode) {
        if (error && error.error && error.error.Message) {
            /** @type {?} */
            var message = error.error.Message;
            if (actionCode) {
                message = actionCode + "操作失败:" + message;
            }
            return message;
        }
    };
    /**
     * 启动流程预解析
     * @param payload API所需的上下文参数
     */
    /**
     * 启动流程预解析
     * @param {?} payload API所需的上下文参数
     * @return {?}
     */
    TaskService.prototype.startProcessSimulate = /**
     * 启动流程预解析
     * @param {?} payload API所需的上下文参数
     * @return {?}
     */
    function (payload) {
        /** @type {?} */
        var url = "/api/runtime/wf/v1.0/processInstances/startProcess-simulation";
        return this.httpSvc.post(url, payload, {});
    };
    /**
     * 启动流程接口
     * @param payload API所需的上下文参数
     */
    /**
     * 启动流程接口
     * @param {?} payload API所需的上下文参数
     * @return {?}
     */
    TaskService.prototype.startProcess = /**
     * 启动流程接口
     * @param {?} payload API所需的上下文参数
     * @return {?}
     */
    function (payload) {
        /** @type {?} */
        var url = "/api/runtime/wf/v1.0/processInstances/startProcess";
        return this.httpSvc.post(url, payload, {});
    };
    /**
     * 取消提交
     * @param param 参数
     */
    /**
     * 取消提交
     * @param {?} param 参数
     * @return {?}
     */
    TaskService.prototype.cancelSubmit = /**
     * 取消提交
     * @param {?} param 参数
     * @return {?}
     */
    function (param) {
        /** @type {?} */
        var url = "/api/runtime/wf/v1.0/processInstances/cancelSubmit";
        return this.httpSvc.post(url, param, {});
    };
    return TaskService;
}());
export { TaskService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaskService.prototype.httpSvc;
}
