// @gsp-svc/file-viewer/esm2015/lib/doc-viewer/doc-viewer.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DocViewerComponent } from './doc-viewer/doc-viewer.component';
import { GspDocumentModule } from '../gsp-document/gsp-document.module';
import { GspTopToolbarModule } from '../gsp-top-toolbar/gsp-top-toolbar.module';
import { PagePreloadService } from '../service/page-preload.service';
import { NavigateService } from '../service/navigate.service';
export class DocViewerModule {
}
DocViewerModule.decorators = [
    { type: NgModule, args: [{
                declarations: [DocViewerComponent],
                imports: [
                    CommonModule,
                    GspDocumentModule,
                    GspTopToolbarModule
                ],
                exports: [DocViewerComponent],
                entryComponents: [DocViewerComponent],
                providers: [PagePreloadService, NavigateService]
            },] }
];
