// @farris/mobile-vue-adapter/esm5/lib/binding-states-mapper.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/binding-states-mapper.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { recomputable } from './recompute';
/**
 * @return {?}
 */
function mapEntityListData() {
    return recomputable((/**
     * @return {?}
     */
    function () {
        return this.$bindingStates.getEntityListData();
    }), 'entityListData');
}
/**
 * @return {?}
 */
function mapCurrentEntityData() {
    return recomputable((/**
     * @return {?}
     */
    function () {
        return this.$bindingStates.getCurrentEntityData();
    }), 'currentEntityData');
}
/**
 * @return {?}
 */
function mapUIState() {
    return recomputable((/**
     * @return {?}
     */
    function () {
        return this.$bindingStates.getUIState();
    }), 'uiState');
}
/**
 * @return {?}
 */
function mapForm() {
    return recomputable((/**
     * @return {?}
     */
    function () {
        return this.$bindingStates.getForm();
    }), 'form');
}
/**
 * @return {?}
 */
function mapStateMachine() {
    return recomputable((/**
     * @return {?}
     */
    function () {
        return this.$bindingStates.getStateMachine();
    }), 'stateMachine');
}
/**
 * @param {?=} extraStates
 * @return {?}
 */
function mapBindingStates(extraStates) {
    /** @type {?} */
    var mappedBindingStates = {
        entityListData: mapEntityListData(),
        currentEntityData: mapCurrentEntityData(),
        uiState: mapUIState(),
        form: mapForm(),
        stateMachine: mapStateMachine()
    };
    /** @type {?} */
    var allMappedStates = Object.assign({}, mappedBindingStates, extraStates);
    return allMappedStates;
}
export { mapBindingStates, mapEntityListData, mapCurrentEntityData, mapUIState, mapForm };
