// @farris/mobile-vue-adapter/esm5/lib/providers-mapper.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/providers-mapper.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BefLookupDataService } from '@farris/mobile-bef';
import { CloudPlusAttachment } from './attachment/attachment-children.service';
import { UploadAttachmentService } from './attachment/upload-attachment.service';
import { LookupMappingService } from './lookup-mapping.service';
import { CloudPlusMapService } from './map/cloud-plus-map.service';
/**
 * 映射Providers
 * @return {?}
 */
function mapProviders() {
    return (/**
     * @return {?}
     */
    function provide() {
        /** @type {?} */
        var viewModel = this.$getViewModel();
        /** @type {?} */
        var lookupDataService = new BefLookupDataService(viewModel.context);
        /** @type {?} */
        var lookupMappingService = new LookupMappingService(viewModel.context);
        // 云上协同附件服务
        /** @type {?} */
        var attachmentService = new CloudPlusAttachment();
        // caf提供附件上传和预览接口调用
        /** @type {?} */
        var uploadAttachmentService = new UploadAttachmentService((/** @type {?} */ (viewModel.context.repository)));
        // 云上协同地图服务
        /** @type {?} */
        var mapService = new CloudPlusMapService();
        /** @type {?} */
        var providers = {
            LookupHttpService: lookupDataService,
            LookupMappingFn: lookupMappingService,
            uploadAttachmentService: uploadAttachmentService,
            attachmentService: attachmentService,
            mapService: mapService
        };
        return providers;
    });
}
export { mapProviders };
