// @gsp-wf/wf-dimension-def/esm2015/lib/entity/dimendef-entity.util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DimensionType, DimensionInputType } from './dimensiondef.entity';
import { Guid } from './guid.util';
class DimenDefEntityUtil {
    constructor() {
    }
    /**
     * @param {?} flowFormId
     * @return {?}
     */
    static newDimendefEntity(flowFormId) {
        /** @type {?} */
        const newDimendef = {
            id: '',
            code: '',
            entryFormID: '',
            dimensionType: DimensionType.BusinessField,
            dimensionCode: '',
            dimensionName: '',
            inputType: DimensionInputType.Help,
            enumList: null,
            smartHelp: '',
            multiSelect: true,
            helpFilter: '',
            isUse: true,
            devLayer: '',
            contentInfo: null,
            sortOrder: 0,
            sysInit: 0,
            remarks: ''
        };
        newDimendef.id = Guid.newGuid();
        newDimendef.entryFormID = flowFormId;
        return newDimendef;
    }
}
export { DimenDefEntityUtil };
