// @edp-pmf/bpmn-model/dist/esm/lib/model/Association.js
import { __extends } from "tslib";
import { Artifact } from "./Artifact";
import { AssociationDirection } from "./AssociationDirection";
import { ExtensionMethod } from "./ExtensionMethod";
var Association = /** @class */ (function (_super) {
    __extends(Association, _super);
    function Association(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.Association";
        _this.AssociationDirection = AssociationDirection.NONE;
        return _this;
    }
    Association.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "SourceRef", this.SourceRef);
        ExtensionMethod.SetAttributeValue(obj, "TargetRef", this.TargetRef);
        return obj;
    };
    Association.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.SourceRef = ExtensionMethod.GetAttributeValue4String(obj, "SourceRef");
        this.TargetRef = ExtensionMethod.GetAttributeValue4String(obj, "TargetRef");
    };
    Association.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
    };
    Association.prototype.NewInstance = function () {
        return new Association(this.Model);
    };
    return Association;
}(Artifact));
export { Association };
