// @edp-pmf/bpmn-model/dist/esm/lib/model/BoundaryEvent.js
import { __extends } from "tslib";
import { BpmnModelHelper } from "./BpmnModelHelper";
import { CatchEvent } from "./CatchEvent";
var BoundaryEvent = /** @class */ (function (_super) {
    __extends(BoundaryEvent, _super);
    function BoundaryEvent(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.BoundaryEvent";
        _this.CancelActivity = true;
        return _this;
    }
    BoundaryEvent.prototype.clone = function () {
        var clone = _super.prototype.clone.call(this);
        clone.AttachedToRef = this.AttachedToRef;
        clone.CancelActivity = this.CancelActivity;
        return clone;
    };
    BoundaryEvent.prototype.NewInstance = function () {
        return new BoundaryEvent(this.Model);
    };
    BoundaryEvent.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.AttachedToRef = null;
        this.CancelActivity = true;
    };
    BoundaryEvent.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "AttachedToRef", this.AttachedToRef);
        BpmnModelHelper.SetAttributeValue(obj, "CancelActivity", this.CancelActivity);
        return obj;
    };
    BoundaryEvent.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.AttachedToRef = obj.AttachedToRef;
        this.CancelActivity = obj.CancelActivity;
    };
    return BoundaryEvent;
}(CatchEvent));
export { BoundaryEvent };
