// @edp-pmf/bpmn-model/dist/esm/lib/model/CatchEvent.js
import { __extends } from "tslib";
import { BpmnModelHelper } from "./BpmnModelHelper";
import { DataOutput } from "./DataOutput";
import { DataOutputAssociation } from "./DataOutputAssociation";
import { ValidateHelper } from "./ValidateHelper";
import { Event } from "./Event";
var CatchEvent = /** @class */ (function (_super) {
    __extends(CatchEvent, _super);
    function CatchEvent(model) {
        var _this = _super.call(this, model) || this;
        _this.dataOutputAssociationDic = {};
        _this.dataOutputDIc = {};
        return _this;
    }
    CatchEvent.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.dataOutputAssociationDic = {};
        this.dataOutputDIc = {};
    };
    CatchEvent.prototype.AddDataOutput = function (item) {
        this.CheckVariableElement(item);
        this.dataOutputDIc[item.Id] = item;
        this.Model.AddHasIdElement(item);
    };
    CatchEvent.prototype.GetDataOutput = function (dataOutputId) {
        if (dataOutputId == null)
            return null;
        if (this.dataOutputDIc.hasOwnProperty("dataOutputId"))
            return this.dataOutputDIc[dataOutputId];
        return null;
    };
    CatchEvent.prototype.GetDataOutputs = function () {
        var dataOutputs = new Array();
        for (var item in this.dataOutputDIc) {
            dataOutputs.push(this.dataOutputDIc[item]);
        }
        return dataOutputs;
    };
    CatchEvent.prototype.RemoveDataOutput = function (dataOutputId) {
        delete this.dataOutputDIc[dataOutputId];
        this.Model.RemoveHasIdElement(dataOutputId);
    };
    CatchEvent.prototype.AddDataOutputAssociation = function (association) {
        ValidateHelper.CheckHasIdElement(association);
        this.dataOutputAssociationDic[association.Id] = association;
        this.Model.AddHasIdElement(association);
    };
    CatchEvent.prototype.GetDataOutputAssociation = function (associationId) {
        if (associationId == null)
            return null;
        if (this.dataOutputAssociationDic.hasOwnProperty("associationId"))
            return this.dataOutputAssociationDic[associationId];
        return null;
    };
    CatchEvent.prototype.GetDataOutputAssociations = function () {
        var dataOutputAssociations = new Array();
        for (var item in this.dataOutputAssociationDic) {
            dataOutputAssociations.push(this.dataOutputAssociationDic[item]);
        }
        return dataOutputAssociations;
    };
    CatchEvent.prototype.RemoveDataOutputAssociation = function (associationId) {
        delete this.dataOutputAssociationDic[associationId];
        this.Model.RemoveHasIdElement(associationId);
    };
    CatchEvent.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        var dataOutputs = new Array();
        for (var item in this.dataOutputDIc) {
            var element = this.dataOutputDIc[item].ConvertToJsonObject();
            dataOutputs.push(element);
        }
        if (dataOutputs.length > 0) {
            BpmnModelHelper.SetAttributeValue(obj, "dataOutputs", dataOutputs);
        }
        var dataOutputAssociations = new Array();
        for (var item in this.dataOutputAssociationDic) {
            var element = this.dataOutputAssociationDic[item].ConvertToJsonObject();
            dataOutputAssociations.push(element);
        }
        if (dataOutputAssociations.length > 0) {
            BpmnModelHelper.SetAttributeValue(obj, "dataOutputAssociations", dataOutputAssociations);
        }
        return obj;
    };
    CatchEvent.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        if (obj.hasOwnProperty("dataOutputs") && Array.isArray(obj.dataOutputs)) {
            for (var _i = 0, _a = obj.dataOutputs; _i < _a.length; _i++) {
                var item = _a[_i];
                var dataOutput = new DataOutput(this.Model);
                dataOutput.LoadFromJsonObject(item);
                this.AddDataOutput(dataOutput);
            }
        }
        if (obj.hasOwnProperty("dataOutputAssociations") && Array.isArray(obj.dataOutputAssociations)) {
            for (var item in obj.dataOutputAssociations) {
                var dataOutputAssociation = new DataOutputAssociation(this.Model);
                dataOutputAssociation.LoadFromJsonObject(item);
                this.AddDataOutputAssociation(dataOutputAssociation);
            }
        }
    };
    CatchEvent.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
        for (var item in this.dataOutputDIc) {
            this.dataOutputDIc[item].Validate(errors);
        }
        for (var item in this.dataOutputAssociationDic) {
            this.dataOutputAssociationDic[item].Validate(errors);
        }
    };
    CatchEvent.prototype.CheckVariableElement = function (variableElement) {
        ValidateHelper.CheckVariableElement(variableElement, null, this, null);
    };
    return CatchEvent;
}(Event));
export { CatchEvent };
