// @edp-pmf/bpmn-model/dist/esm/lib/model/CompensateEventDefinition.js
import { __extends } from "tslib";
import { EventDefinition } from "./EventDefinition";
import { ExtensionMethod } from "./ExtensionMethod";
var CompensateEventDefinition = /** @class */ (function (_super) {
    __extends(CompensateEventDefinition, _super);
    function CompensateEventDefinition(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.Association";
        _this.WaitForCompletion = true;
        return _this;
    }
    CompensateEventDefinition.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "ActivityRef", this.ActivityRef);
        ExtensionMethod.SetAttributeValue(obj, "WaitForCompletion", this.WaitForCompletion);
        return obj;
    };
    CompensateEventDefinition.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.ActivityRef = ExtensionMethod.GetAttributeValue4String(obj, "ActivityRef");
        this.WaitForCompletion = ExtensionMethod.GetAttributeValue4Boolean(obj, "WaitForCompletion");
    };
    CompensateEventDefinition.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
    };
    CompensateEventDefinition.prototype.NewInstance = function () {
        var clone = new CompensateEventDefinition(this.Model);
        clone.ActivityRef = this.ActivityRef;
        clone.WaitForCompletion = this.WaitForCompletion;
        return clone;
    };
    return CompensateEventDefinition;
}(EventDefinition));
export { CompensateEventDefinition };
