// @edp-pmf/bpmn-model/dist/esm/lib/model/ExclusiveGateway.js
import { __extends } from "tslib";
import { BpmnModelHelper } from "./BpmnModelHelper";
import { Gateway } from "./Gateway";
import { GatewayType } from "./GatewayType";
var ExclusiveGateway = /** @class */ (function (_super) {
    __extends(ExclusiveGateway, _super);
    function ExclusiveGateway(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.ExclusiveGateway";
        _this.IncomingGatewayType = GatewayType.Exclusive;
        _this.OutgoingGatewayType = GatewayType.Exclusive;
        return _this;
    }
    ExclusiveGateway.prototype.NewInstance = function () {
        return new ExclusiveGateway(this.Model);
    };
    ExclusiveGateway.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Default = null;
    };
    ExclusiveGateway.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "Default", this.Default);
        return obj;
    };
    ExclusiveGateway.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Default = obj.Default;
    };
    return ExclusiveGateway;
}(Gateway));
export { ExclusiveGateway };
