// @edp-pmf/bpmn-model/dist/esm/lib/model/InclusiveGateway.js
import { __extends } from "tslib";
import { BpmnModelHelper } from "./BpmnModelHelper";
import { Gateway } from "./Gateway";
import { GatewayType } from "./GatewayType";
var InclusiveGateway = /** @class */ (function (_super) {
    __extends(InclusiveGateway, _super);
    function InclusiveGateway(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.InclusiveGateway";
        _this.IncomingGatewayType = GatewayType.Inclusive;
        _this.OutgoingGatewayType = GatewayType.Inclusive;
        return _this;
    }
    InclusiveGateway.prototype.NewInstance = function () {
        return new InclusiveGateway(this.Model);
    };
    InclusiveGateway.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Default = null;
    };
    InclusiveGateway.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "Default", this.Default);
        return obj;
    };
    InclusiveGateway.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Default = obj.Default;
    };
    return InclusiveGateway;
}(Gateway));
export { InclusiveGateway };
