// @edp-pmf/bpmn-model/dist/esm/lib/model/MergeActivity.js
import { __assign, __extends } from "tslib";
import { ServiceTask } from "./ServiceTask";
var MergeActivity = /** @class */ (function (_super) {
    __extends(MergeActivity, _super);
    function MergeActivity(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = MergeActivity.Clr_Type_ID;
        return _this;
    }
    MergeActivity.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    MergeActivity.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        return obj;
    };
    MergeActivity.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
    };
    MergeActivity.prototype.NewInstance = function () {
        return new MergeActivity(this.Model);
    };
    MergeActivity.prototype.clone = function () {
        var c = _super.prototype.clone.call(this);
        c.Name = this.Name;
        c.NameLanguage = this.NameLanguage && __assign({}, this.NameLanguage);
        return c;
    };
    MergeActivity.Clr_Type_ID = "bpmn.MergeActivity";
    MergeActivity.MEXTAG = 'bpmn.MergeActivity';
    return MergeActivity;
}(ServiceTask));
export { MergeActivity };
