// @edp-pmf/bpmn-model/dist/esm/lib/model/SequenceFlow.js
import { __extends } from "tslib";
import { BpmnModelHelper } from "./BpmnModelHelper";
import { Expression } from "./Expression";
import { FlowElement } from "./FlowElement";
import { ExecutionListener } from "./mex/ExecutionListener";
var SequenceFlow = /** @class */ (function (_super) {
    __extends(SequenceFlow, _super);
    function SequenceFlow(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.SequenceFlow";
        _this.executionListeners = new Array();
        _this.MEXTAG = "bpmn.SequenceFlow";
        return _this;
    }
    SequenceFlow.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.ConditionExpression = null;
        this.SkipExpression = null;
        this.SourceRef = null;
        this.TargetRef = null;
        this.executionListeners = new Array();
    };
    SequenceFlow.prototype.NewInstance = function () {
        return new SequenceFlow(this.Model);
    };
    SequenceFlow.prototype.GetExecutionListeners = function () {
        return this.executionListeners;
    };
    SequenceFlow.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "SkipExpression", this.SkipExpression);
        BpmnModelHelper.SetAttributeValue(obj, "SourceRef", this.SourceRef);
        BpmnModelHelper.SetAttributeValue(obj, "TargetRef", this.TargetRef);
        if (this.ConditionExpression != null) {
            BpmnModelHelper.SetAttributeValue(obj, "ConditionExpression", this.ConditionExpression.ConvertToJsonObject());
        }
        return obj;
    };
    SequenceFlow.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.SkipExpression = obj.SkipExpression;
        this.SourceRef = obj.SourceRef;
        this.TargetRef = obj.TargetRef;
        if (obj.hasOwnProperty("ConditionExpression")) {
            this.ConditionExpression = new Expression(this.Model);
            this.ConditionExpression.LoadFromJsonObject(obj.ConditionExpression);
        }
    };
    SequenceFlow.prototype.GetExtensionElements4ConvertTo = function () {
        var list = _super.prototype.GetExtensionElements4ConvertTo.call(this);
        for (var _i = 0, _a = this.executionListeners; _i < _a.length; _i++) {
            var item = _a[_i];
            item.MexTag = this.MEXTAG;
            list.push(item);
        }
        return list;
    };
    SequenceFlow.prototype.SetExtensionElements4LoadFrom = function (extensionElements) {
        _super.prototype.SetExtensionElements4LoadFrom.call(this, extensionElements);
        for (var _i = 0, extensionElements_1 = extensionElements; _i < extensionElements_1.length; _i++) {
            var extensionElement = extensionElements_1[_i];
            if (extensionElement instanceof ExecutionListener && extensionElement.MexTag == this.MEXTAG) {
                this.executionListeners.push(extensionElement);
            }
        }
    };
    SequenceFlow.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
        for (var _i = 0, _a = this.executionListeners; _i < _a.length; _i++) {
            var item = _a[_i];
            item.Validate(errors);
        }
    };
    return SequenceFlow;
}(FlowElement));
export { SequenceFlow };
