// @edp-pmf/bpmn-model/dist/esm/lib/model/SignalEventDefinition.js
import { __extends } from "tslib";
import { EventDefinition } from "./EventDefinition";
import { ExtensionMethod } from "./ExtensionMethod";
var SignalEventDefinition = /** @class */ (function (_super) {
    __extends(SignalEventDefinition, _super);
    function SignalEventDefinition(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.SignalEventDefinition";
        return _this;
    }
    SignalEventDefinition.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "ClrTypeID", this.ClrTypeID);
        ExtensionMethod.SetAttributeValue(obj, "SignalRef", this.SignalRef);
        ExtensionMethod.SetAttributeValue(obj, "SignalExpression", this.SignalExpression);
        ExtensionMethod.SetAttributeValue(obj, "Async", this.Async);
        return obj;
    };
    SignalEventDefinition.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.SignalRef = ExtensionMethod.GetAttributeValue4String(obj, "SignalRef");
        this.SignalExpression = ExtensionMethod.GetAttributeValue4String(obj, "SignalExpression");
        this.Async = ExtensionMethod.GetAttributeValue4Boolean(obj, "Async");
    };
    SignalEventDefinition.prototype.NewInstance = function () {
        var clone = new SignalEventDefinition(this.Model);
        clone.SignalRef = this.SignalRef;
        clone.SignalExpression = this.SignalExpression;
        clone.Async = this.Async;
        return clone;
    };
    return SignalEventDefinition;
}(EventDefinition));
export { SignalEventDefinition };
