// @gspwidget/util/esm2015/lib/arr.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class ArrUtil {
    /**
     * @param {?} arr
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    static SwapByIndex(arr, i, j) {
        /** @type {?} */
        const tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }
    /**
     * @param {?} arr
     * @param {?} newItem
     * @return {?}
     */
    static Add(arr, newItem) {
        arr.push(newItem);
    }
    /**
     * @param {?} arr
     * @param {?} item
     * @return {?}
     */
    static Up(arr, item) {
        /** @type {?} */
        const i = arr.indexOf(item);
        if (i < 1)
            return;
        ArrUtil.SwapByIndex(arr, i, i - 1);
    }
    /**
     * @param {?} arr
     * @param {?} item
     * @return {?}
     */
    static Down(arr, item) {
        /** @type {?} */
        const i = arr.indexOf(item);
        if (i > arr.length - 2)
            return;
        ArrUtil.SwapByIndex(arr, i, i + 1);
    }
    /**
     * @param {?} arr
     * @param {?} item
     * @return {?}
     */
    static Remove(arr, item) {
        /** @type {?} */
        const i = arr.indexOf(item);
        if (i === -1)
            return;
        arr.splice(i, 1);
        // arr = arr.filter(i => item!==i)
    }
    /**
     * @param {?} arr
     * @return {?}
     */
    static Unique(arr) {
        return Array.from(new Set(arr));
    }
}
