// @gspwidget/util/esm2015/lib/colors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 部件系统默认色系。
 * 直接用的缺陷是无法拿到 home.config.json 中定制的色系值，
 * 尽量用 colors
 */
export class PlThemeColors {
    /**
     * @param {?} pure
     * @param {?} gradient
     * @return {?}
     */
    static SetColors(pure, gradient) {
        this.SetPure(pure);
        this.SetGradient(gradient);
    }
    /**
     * @param {?} pure
     * @return {?}
     */
    static SetPure(pure) {
        if (pure && pure.length > 0)
            this.pure = pure;
    }
    /**
     * @param {?} gradient
     * @return {?}
     */
    static SetGradient(gradient) {
        if (gradient && gradient.length > 0)
            this.gradient = gradient;
    }
}
PlThemeColors.pure = [
    '#F5B805', '#FF9300', '#EC7658', '#897AFB', '#5E6EEB',
    '#4D97FF', '#01BBC7', '#3DDB88', '#A4EA4D', '#93A3D3',
    '#000000', '#ffffff'
];
PlThemeColors.gradient = [
    '#FBDD0C,#F5B805',
    '#FFB046,#FF9300',
    '#FF9579,#EC7658',
    '#8B99FF,#897AFB',
    '#7F8DFF,#5E6EEB',
    '#79B1FF,#4D97FF',
    '#00DBE9,#01BBC7',
    '#46FA9B,#3DDB88',
    '#C1FF73,#A4EA4D',
    '#B5C7FD,#93A3D3',
];
if (false) {
    /** @type {?} */
    PlThemeColors.pure;
    /** @type {?} */
    PlThemeColors.gradient;
}
