// @gspwidget/util/esm2015/lib/const.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 通过 ls 中 LS_LOAD_LOCAL_ASSET 标记判断是否加载 localhost/assets。仅用于部件系统开发人员，只能手动修改 ls 中的标记值。
 * @type {?}
 */
export const LS_LOAD_LOCAL_ASSET = 'LS_LOAD_LOCAL_ASSET';
/** @type {?} */
const loadLocalAsset = localStorage.getItem(LS_LOAD_LOCAL_ASSET);
/** @type {?} */
export const IGIX_HOME_PATH = loadLocalAsset ? `` : `/platform/runtime/sys/web/home/`;
/** @type {?} */
export const BASE_URL = `/api/runtime/epp/v1.0/`;
/**
 * @param {?} node
 * @return {?}
 */
export function addBaseUrl(node) {
    Object.keys(node).forEach((/**
     * @param {?} key
     * @return {?}
     */
    key => {
        if (typeof node[key] === 'string') {
            node[key] = BASE_URL + node[key];
        }
        else {
            addBaseUrl(node[key]);
        }
    }));
}
// const DATA_PREFIX = `/jersey`
/**
 * 数据源相关的 base url, 为了防止跟其他的不同. 若相同, 可删掉
 * @type {?}
 */
export const DATA_PREFIX = `data`;
/** @type {?} */
export const FILTERS_WIDGET_NAME = 'filters';
/** @type {?} */
export const FILTERS_WIDGET_FILTERS_PROP_NAME = 'filters';
/** @type {?} */
export const FILTERS_WIDGET_PROP_NAME = 'change';
/** @type {?} */
export const DESIGNER_CONFIG = {
    PAGE_DESIGNER_ID: `Epp-pagedesigner`,
    PAGE_RUNTIME_ID: `Epp-pageruntime`
};
/** @type {?} */
export const dimensionTypeConfig = {
    text: {
        list: ['STRING', 'TEXT'],
        color: '#2196F3',
        iconName: 'texture'
    },
    number: {
        // TODO: integer
        list: ['SHORT', 'LONG', 'FLOAT', 'DOUBLE', 'DECIMAL', 'INTEGER', 'INTETER'],
        color: '#e91e63',
        iconName: 'device_hub'
    },
    date: {
        list: ['DATETIME'],
        color: '#ff9800',
        iconName: 'access_time'
    },
    byte: {
        list: ['BYTE', 'BYTES'],
        color: '#4caf50',
        iconName: 'texture'
    },
    boolean: {
        list: ['BOOLEAN', 'UNKNOWN'],
        color: '#3f51b5',
        iconName: 'invert_colors'
    },
    unknown: {
        list: ['UNKNOWN'],
        color: '#3f51b5',
        iconName: 'help'
    }
};
