// @gspwidget/util/esm2015/lib/rtf-util.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
/** @type {?} */
const WINDOW_RTF_KEY = 'gspframeworkService';
/**
 * @record
 */
export function RtfTabChangeEvent() { }
if (false) {
    /** @type {?} */
    RtfTabChangeEvent.prototype.appEntrance;
    /** @type {?} */
    RtfTabChangeEvent.prototype.appId;
    /** @type {?} */
    RtfTabChangeEvent.prototype.appType;
    /** @type {?} */
    RtfTabChangeEvent.prototype.funcId;
    /** @type {?} */
    RtfTabChangeEvent.prototype.tabId;
    /** @type {?} */
    RtfTabChangeEvent.prototype.tabName;
}
export class RtfUtilService {
    /**
     * @param {?} translate
     */
    constructor(translate) {
        this.translate = translate;
    }
    // TODO: queryParamArray 是个 array 就行
    /**
     * @param {?} funcId
     * @param {?=} queryParamArray
     * @param {?=} tabId
     * @param {?=} isReload
     * @return {?}
     */
    openMenu(funcId, queryParamArray, tabId, isReload) {
        ((/** @type {?} */ (window))).gspframeworkService.rtf.func.openMenu((/** @type {?} */ ({
            appType: 'menu',
            funcId, tabId, isReload,
            appId: '', appEntrance: '',
            queryStringParams: new Map(queryParamArray)
        })));
    }
    /**
     * @param {?} funcId
     * @param {?=} tabId
     * @return {?}
     */
    closeMenu(funcId, tabId) {
        ((/** @type {?} */ (window))).gspframeworkService.rtf.func.close((/** @type {?} */ ({
            appType: 'menu',
            funcId, tabId,
            appId: '', appEntrance: ''
        })));
    }
    /**
     * @param {?} appId
     * @return {?}
     */
    openApp(appId) {
        ((/** @type {?} */ (window))).gspframeworkService.rtf.func.openMenu((/** @type {?} */ ({
            appType: 'app',
            funcId: '',
            appId,
            appEntrance: '',
        })));
    }
    /**
     * @return {?}
     */
    openUserInfoMenu() {
        const { rtf } = window[WINDOW_RTF_KEY];
        /** @type {?} */
        const frmEventSvc = rtf.frmEvent;
        /** @type {?} */
        const sessionSvc = rtf.adapter.adatperSerice().session;
        this.translate.get('userInfo').subscribe((/**
         * @param {?} txt
         * @return {?}
         */
        txt => {
            frmEventSvc.eventFire(`farrisapp-click`, {
                active: false,
                code: 'userInfo',
                src: '/platform/runtime/sys/web/userinfo/index.html',
                sessionid: sessionSvc.getUserSessionId(),
                id: 'userinfo',
                appType: '',
                FuncName: txt
            });
        }));
    }
    /**
     * @param {?} route
     * @return {?}
     */
    closeCurRtfTab(route) {
        const { tabId, funcId } = route.snapshot.queryParams;
        this.closeMenu(funcId, tabId);
    }
    /**
     * @param {?} route
     * @param {?=} condition
     * @return {?}
     */
    onRtfTabChangeToMe(route, condition) {
        /** @type {?} */
        const onRtfTabChangeToMe$ = new EventEmitter();
        const { tabId: curTabId } = route.snapshot.queryParams;
        onRtfTabChange().subscribe((/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            if (!condition)
                condition = (/**
                 * @param {?} pToTabId
                 * @param {?} pCurTabId
                 * @return {?}
                 */
                (pToTabId, pCurTabId) => pToTabId === pCurTabId);
            if (condition(e.tabId, curTabId)) {
                onRtfTabChangeToMe$.emit(e);
            }
        }));
        return onRtfTabChangeToMe$;
    }
}
RtfUtilService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RtfUtilService.ctorParameters = () => [
    { type: TranslateService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    RtfUtilService.prototype.translate;
}
/** @type {?} */
const onRtfTabChange$ = new EventEmitter();
/** @type {?} */
let funcSwitchEventRegistered = false
/** 用函数是为了延迟注册事件，否则 rtf service 可能还没挂在 window 上 */
;
/**
 * 用函数是为了延迟注册事件，否则 rtf service 可能还没挂在 window 上
 * @return {?}
 */
export function onRtfTabChange() {
    if (!funcSwitchEventRegistered && window[WINDOW_RTF_KEY]) {
        funcSwitchEventRegistered = true;
        window[WINDOW_RTF_KEY].rtf.frmEvent.eventListener('funcSwitchEvent', (/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            onRtfTabChange$.emit(e);
        }));
    }
    return onRtfTabChange$;
}
