// @farris/extend-scroll-collapsible-area/esm2015/lib/scroll-area.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/scroll-area.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef } from '@angular/core';
import { fromEvent } from 'rxjs';
import { debounceTime } from 'rxjs/operators';
export class FEScrollAreaDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        // 记录滚动的值
        this.scrollAreaTop = 0;
        // 记录值
        this.threshold = 0;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.bindScrollEvent();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.scrollSub.unsubscribe();
        this.scrollSub = null;
    }
    // 追加关联
    /**
     * @param {?} cInstance
     * @return {?}
     */
    addCollapseInstance(cInstance) {
        this.relatedCollapseInstance = cInstance;
    }
    // 绑定事件
    /**
     * @return {?}
     */
    bindScrollEvent() {
        this.scrollSub = fromEvent(this.el.nativeElement, 'scroll').pipe(debounceTime(20)).subscribe((/**
         * @return {?}
         */
        () => {
            //判断滚动的方向
            this.scrollDirection();
        }));
    }
    /**
     * @return {?}
     */
    scrollDirection() {
        /** @type {?} */
        let afterScrollTop = this.el.nativeElement.scrollTop;
        /** @type {?} */
        let delta = afterScrollTop - this.scrollAreaTop;
        // 没有变化
        if (delta === 0)
            return false;
        this.scrollAreaTop = afterScrollTop;
        this.judgeThreshold(delta > 0 ? "down" : "up");
    }
    /**
     * @private
     * @param {?} direction
     * @return {?}
     */
    judgeThreshold(direction) {
        this.relatedCollapseInstance.updateHeight(this.scrollAreaTop, direction);
    }
}
FEScrollAreaDirective.decorators = [
    { type: Directive, args: [{
                selector: '[fe-scroll-area]'
            },] }
];
/** @nocollapse */
FEScrollAreaDirective.ctorParameters = () => [
    { type: ElementRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    FEScrollAreaDirective.prototype.scrollSub;
    /**
     * @type {?}
     * @private
     */
    FEScrollAreaDirective.prototype.relatedCollapseInstance;
    /**
     * @type {?}
     * @private
     */
    FEScrollAreaDirective.prototype.scrollAreaTop;
    /**
     * @type {?}
     * @private
     */
    FEScrollAreaDirective.prototype.threshold;
    /**
     * @type {?}
     * @private
     */
    FEScrollAreaDirective.prototype.el;
}
