// @farris/extend-scroll-collapsible-area/esm2015/lib/scroll-collapsible-area.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/scroll-collapsible-area.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, ContentChild } from '@angular/core';
import { FECollapsibleAreaComponent } from './collapsible-area/collapsible-area.component';
import { FEScrollAreaDirective } from './scroll-area.directive';
export class FEScrollCollapsibleAreaDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.scrollArea.addCollapseInstance(this.collapseArea);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // setTimeout(()=>{
        //   debugger;
        //   this.scrollArea.addCollapseInstance(this.collapseArea);
        // },300);
    }
}
FEScrollCollapsibleAreaDirective.decorators = [
    { type: Directive, args: [{
                selector: '[fe-scroll-collapsible]'
            },] }
];
/** @nocollapse */
FEScrollCollapsibleAreaDirective.ctorParameters = () => [
    { type: ElementRef }
];
FEScrollCollapsibleAreaDirective.propDecorators = {
    collapseArea: [{ type: ContentChild, args: [FECollapsibleAreaComponent,] }],
    scrollArea: [{ type: ContentChild, args: [FEScrollAreaDirective,] }]
};
if (false) {
    /** @type {?} */
    FEScrollCollapsibleAreaDirective.prototype.collapseArea;
    /** @type {?} */
    FEScrollCollapsibleAreaDirective.prototype.scrollArea;
    /**
     * @type {?}
     * @private
     */
    FEScrollCollapsibleAreaDirective.prototype.el;
}
