// @progress/kendo-angular-dialog/dist/es2015/dialog/dialog-titlebar.component.js
import { Component, EventEmitter, HostBinding, Output, Input, ElementRef } from '@angular/core';
/**
 * @hidden
 */
export class DialogTitleBarComponent {
    constructor(hostElement) {
        this.hostElement = hostElement;
        this.close = new EventEmitter();
    }
    get className() {
        return true;
    }
    ngAfterViewChecked() {
        const element = this.hostElement.nativeElement;
        element.setAttribute('id', this.id);
    }
    onCloseClick(e) {
        e.preventDefault();
        this.close.emit();
    }
}
DialogTitleBarComponent.decorators = [
    { type: Component, args: [{
                selector: 'kendo-dialog-titlebar',
                template: `
    <div class="k-window-title k-dialog-title">
      <ng-content></ng-content>
    </div>
    <div class="k-window-actions k-dialog-actions">
      <a href="#" role="button"
         aria-label="Close"
         class="k-button k-bare k-button-icon k-window-action k-dialog-action k-dialog-close"
         (click)="onCloseClick($event)">
        <span class="k-icon k-i-x"></span>
      </a>
    </div>
  `
            },] },
];
/** @nocollapse */
DialogTitleBarComponent.ctorParameters = () => [
    { type: ElementRef, },
];
DialogTitleBarComponent.propDecorators = {
    'close': [{ type: Output },],
    'id': [{ type: Input },],
    'className': [{ type: HostBinding, args: ['class.k-window-titlebar',] }, { type: HostBinding, args: ['class.k-dialog-titlebar',] }, { type: HostBinding, args: ['class.k-header',] },],
};
