// @progress/kendo-angular-dialog/dist/es2015/dialog/dialog-actions.component.js
import { Component, EventEmitter, HostBinding, TemplateRef, Input, Output, ElementRef } from '@angular/core';
/**
 * Specifies the action buttons of the Dialog
 * ([see example]({% slug actionbuttons_dialog %})).
 */
export class DialogActionsComponent {
    constructor(el) {
        this.el = el;
        /**
         * Fires when the user clicks the actions that are specified by the action buttons.
         */
        this.action = new EventEmitter();
    }
    get className() {
        return true;
    }
    /**
     * @hidden
     */
    actionTemplate() {
        return this.actions instanceof TemplateRef;
    }
    /**
     * @hidden
     */
    onButtonClick(action, _e) {
        this.action.emit(action);
    }
    /**
     * @hidden
     */
    buttonClass(action) {
        const classes = ["k-button"];
        if (action.primary) {
            classes.push("k-primary");
        }
        return classes.join(" ");
    }
}
DialogActionsComponent.decorators = [
    { type: Component, args: [{
                selector: 'kendo-dialog-actions',
                template: `
    <ng-content *ngIf="!actions"></ng-content>
    <ng-container *ngIf="!actionTemplate()">
      <button
        [ngClass]="buttonClass(action)"
        (click)="onButtonClick(action, $event)"
        *ngFor="let action of actions">
        {{ action.text }}
      </button>
    </ng-container>
    <ng-template [ngTemplateOutlet]="actions" *ngIf="actionTemplate()"></ng-template>
  `
            },] },
];
/** @nocollapse */
DialogActionsComponent.ctorParameters = () => [
    { type: ElementRef, },
];
DialogActionsComponent.propDecorators = {
    'actions': [{ type: Input },],
    'action': [{ type: Output },],
    'className': [{ type: HostBinding, args: ['class.k-dialog-buttongroup',] }, { type: HostBinding, args: ['class.k-dialog-button-layout-stretched',] },],
};
