// @gsp-bef/gsp-vo-metadata/esm5/lib/view-model/src/Enum/MappingType.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 映射类型枚举
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
var MappingType = {
    /**
     * BE
     */
    BizEntity: 'BizEntity',
    /**
     * DM
     */
    DataModel: 'DataModel',
    /**
     * BE节点
     */
    BEObject: 'BEObject',
    /**
     * GspDataModel节点
     */
    DMObject: 'DMObject',
    /**
     * 节点元素
     */
    Element: 'Element',
    /**
     * BE操作
     */
    BizOperation: 'BizOperation',
    /**
     * 构件方法
     */
    ComponentMethod: 'ComponentMethod',
    /**
     * 数据对象
     */
    DataObject: 'DataObject',
    /**
     * 数据对象列
     */
    GSPColumn: 'GSPColumn',
};
export { MappingType };
/** @type {?} */
export var MappingTypes = [
    { text: '业务实体', value: MappingType.BizEntity },
    { text: '业务实体节点', value: MappingType.BEObject },
    { text: '业务实体字段', value: MappingType.Element },
    { text: '业务实体操作', value: MappingType.BizOperation },
    { text: '构件方法', value: MappingType.ComponentMethod },
];
