// @gsp-wf/wf-process-model/esm5/lib/model/extension/JumpScope.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionElement } from "@edp-pmf/bpmn-model";
import { JumpScopeCategory } from "./JumpScopeCategory";
import { JumpScopeDefaultType } from "./JumpScopeDefaultType";
var JumpScope = /** @class */ (function (_super) {
    tslib_1.__extends(JumpScope, _super);
    function JumpScope(model, category, DefaultType) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = JumpScope.Clr_Type_ID;
        _this.Category = category || JumpScopeCategory.SpecifiedSet;
        _this.DefaultType = DefaultType || JumpScopeDefaultType.FirstNode;
        _this.DestinationActivities = new Array();
        _this.DefaultActivity = '';
        return _this;
    }
    /**
     * @return {?}
     */
    JumpScope.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new JumpScope(this.Model);
    };
    /**
     * @protected
     * @return {?}
     */
    JumpScope.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Category = JumpScopeCategory.Global;
        this.DefaultType = JumpScopeDefaultType.FirstNode;
        this.DestinationActivities = new Array();
        this.DefaultActivity = '';
    };
    /**
     * @return {?}
     */
    JumpScope.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "Category", this.Category);
        BpmnModelHelper.SetAttributeValue(obj, "DefaultType", this.DefaultType);
        BpmnModelHelper.SetAttributeValue(obj, "DestinationActivities", this.DestinationActivities);
        BpmnModelHelper.SetAttributeValue(obj, "DefaultActivity", this.DefaultActivity);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    JumpScope.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Category = obj.Category;
        this.DefaultType = obj.DefaultType;
        this.DestinationActivities = obj.DestinationActivities;
        this.DefaultActivity = obj.DefaultActivity;
    };
    /**
     * @return {?}
     */
    JumpScope.prototype.clone = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        /** @type {?} */
        var jumpScope = new JumpScope(this.Model);
        jumpScope.Category = this.Category;
        jumpScope.DefaultType = this.DefaultType;
        jumpScope.DefaultActivity = this.DefaultActivity;
        jumpScope.DestinationActivities = new Array();
        if (this.DestinationActivities && this.DestinationActivities.length > 0) {
            try {
                for (var _b = tslib_1.__values(this.DestinationActivities), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var d = _c.value;
                    jumpScope.DestinationActivities.push(d);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        return jumpScope;
    };
    JumpScope.Clr_Type_ID = "wf.JumpScope";
    return JumpScope;
}(ExtensionElement));
export { JumpScope };
if (false) {
    /** @type {?} */
    JumpScope.Clr_Type_ID;
    /** @type {?} */
    JumpScope.prototype.ClrTypeID;
    /** @type {?} */
    JumpScope.prototype.Category;
    /** @type {?} */
    JumpScope.prototype.DefaultType;
    /** @type {?} */
    JumpScope.prototype.DestinationActivities;
    /** @type {?} */
    JumpScope.prototype.DefaultActivity;
}
BpmnModelHelper.RegisterElementType(new JumpScope(new BpmnModel()));
