// @gsp-wf/wf-process-model/esm5/lib/model/extension/Notification.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelHelper, ExtensionElement } from "@edp-pmf/bpmn-model";
import { ExecutionType } from "../misc/ExecutionType";
var Notification = /** @class */ (function (_super) {
    tslib_1.__extends(Notification, _super);
    function Notification(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = Notification.Clr_Type_ID;
        /// <summary>消息的发送方式：同步或者异步</summary>
        _this.Execution = ExecutionType.SYNCHRONOUS;
        return _this;
    }
    /**
     * @return {?}
     */
    Notification.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new Notification(this.Model);
    };
    /**
     * @protected
     * @return {?}
     */
    Notification.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    Notification.prototype.ToString = /**
     * @return {?}
     */
    function () {
        return "Message[name='" + this.Name + "'" + ", content='" + this.Header + "\r\n" + this.Body + "\r\n" + this.Footer + "'" + ",receiver=" + this.Receiver + "]";
    };
    /**
     * @return {?}
     */
    Notification.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, 'Id', this.Id);
        BpmnModelHelper.SetAttributeValue(obj, 'Name', this.Name);
        BpmnModelHelper.SetAttributeValue(obj, 'Time', this.Time);
        BpmnModelHelper.SetAttributeValue(obj, 'Condition', this.Condition);
        BpmnModelHelper.SetAttributeValue(obj, 'SenderName', this.SenderName);
        BpmnModelHelper.SetAttributeValue(obj, 'Receiver', this.Receiver);
        BpmnModelHelper.SetAttributeValue(obj, 'Way', this.Way);
        BpmnModelHelper.SetAttributeValue(obj, 'Header', this.Header);
        BpmnModelHelper.SetAttributeValue(obj, 'Body', this.Body);
        BpmnModelHelper.SetAttributeValue(obj, 'Footer', this.Footer);
        BpmnModelHelper.SetAttributeValue(obj, 'SendTempAssMessage', this.SendTempAssMessage);
        BpmnModelHelper.SetAttributeValue(obj, 'Execution', this.Execution);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    Notification.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Id = obj.Id;
        this.Name = obj.Name;
        this.Time = (/** @type {?} */ (obj.Time));
        this.Condition = obj.Condition;
        this.SenderName = obj.SenderName;
        this.Receiver = obj.Receiver;
        this.Way = obj.Way;
        this.Header = obj.Header;
        this.Body = obj.Body;
        this.Footer = obj.Footer;
        this.SendTempAssMessage = obj.SendTempAssMessage;
        this.Execution = (/** @type {?} */ (obj.Execution));
    };
    /**
     * @return {?}
     */
    Notification.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var notification = new Notification(this.Model);
        return notification;
    };
    Notification.Clr_Type_ID = "wf.Notification";
    return Notification;
}(ExtensionElement));
export { Notification };
if (false) {
    /** @type {?} */
    Notification.Clr_Type_ID;
    /** @type {?} */
    Notification.prototype.ClrTypeID;
    /** @type {?} */
    Notification.prototype.Id;
    /** @type {?} */
    Notification.prototype.Name;
    /** @type {?} */
    Notification.prototype.Time;
    /** @type {?} */
    Notification.prototype.Condition;
    /** @type {?} */
    Notification.prototype.SenderName;
    /** @type {?} */
    Notification.prototype.Receiver;
    /** @type {?} */
    Notification.prototype.Way;
    /** @type {?} */
    Notification.prototype.Header;
    /** @type {?} */
    Notification.prototype.Body;
    /** @type {?} */
    Notification.prototype.Footer;
    /** @type {?} */
    Notification.prototype.SendTempAssMessage;
    /** @type {?} */
    Notification.prototype.Execution;
}
/** @enum {number} */
var NotifyTime = {
    ///<summary>流程实例预警启动事件</summary>
    PROCESSINSTANCE_ALARM: 0,
    ///<summary>流程实例超时事件</summary> 
    PROCESSINSTANCE_TIMEOUT: 1,
    /// <summary> 过程实例已经异常退出事件 </summary>
    PROCESSINSTANCE_ABORT: 2,
    /// <summary> 过程实例已经完成事件 </summary>  
    PROCESSINSTANCE_COMPLETE: 3,
    /// <summary> 过程实例已经创建事件 </summary>
    PROCESSINSTANCE_CREATE: 4,
    /// <summary> 过程实例已经删除事件 </summary>
    PROCESSINSTANCE_DELETE: 5,
    /// <summary> 过程实例已经暂停事件 </summary>
    PROCESSINSTANCE_RESUME: 6,
    /// <summary> 过程实例已经启动事件 </summary>
    PROCESSINSTANCE_START: 7,
    /// <summary> 过程实例已经挂起事件 </summary>
    PROCESSINSTANCE_SUSPEND: 8,
    /// <summary> 过程实例已经中止事件 </summary>
    PROCESSINSTANCE_TERMINATE: 9,
    /// <summary> 活动实例已经异常退出事件 </summary>
    ACTIVITY_ABORTED: 10,
    /// <summary> 活动实例已经完成事件 </summary>
    ACTIVITY_COMPLETED: 11,
    /// <summary> 活动实例已经创建事件 </summary>
    ACTIVITY_CREATED: 12,
    /// <summary> 活动实例已经暂停事件 </summary>
    ACTIVITY_RESUMED: 13,
    /// <summary> 活动实例已经启动事件 </summary>
    ACTIVITY_STARTED: 14,
    /// <summary> 活动实例已经停止事件 </summary>
    ACTIVITY_STOPPED: 15,
    /// <summary> 活动实例已经挂起事件 </summary>
    ACTIVITY_SUSPENDED: 16,
    /// <summary> 活动实例已经中止事件 </summary>
    ACTIVITY_TERMINATED: 17,
    /// <summary>活动实例已经取回</summary>
    ACTIVITY_RETRIEVED: 18,
    /// <summary>活动实例从该活动取回</summary>
    ACTIVITY_RETRIEVED_FROM: 19,
    /// <summary>活动实例回退</summary>
    ACTIVITY_ROLLBACK: 20,
    /// <summary>活动实例回退到该活动 </summary>
    ACTIVITY_ROLLBACK_TO: 21,
    /// <summary> 工作项已经异常退出事件 </summary>
    WORKITEM_ABORTED: 22,
    /// <summary> 工作项已经分配事件 </summary>
    WORKITEM_ASSIGNED: 23,
    /// <summary> 工作项已经完成事件 </summary>
    WORKITEM_COMPLETED: 24,
    /// <summary> 工作项已经创建事件 </summary>
    WORKITEM_CREATED: 25,
    /// <summary> 工作项已经暂停事件 </summary>
    WORKITEM_RESUMED: 26,
    /// <summary> 工作项已经启动事件 </summary>
    WORKITEM_STARTED: 27,
    /// <summary> 工作项已经停止事件 </summary>
    WORKITEM_STOPPED: 28,
    /// <summary> 工作项已经挂起事件 </summary>
    WORKITEM_SUSPENDED: 29,
    /// <summary> 工作项已经终止事件 </summary>
    WORKITEM_TERMINATED: 30,
    /// <summary> 工作项预警启动事件 </summary>
    WORKITEM_ALARM: 31,
    /// <summary> 工作项超时事件</summary>
    WORKITEM_TIMEOUT: 32,
    /// <summary> 未定义</summary>
    UNDEFINITION: 33,
};
export { NotifyTime };
NotifyTime[NotifyTime.PROCESSINSTANCE_ALARM] = 'PROCESSINSTANCE_ALARM';
NotifyTime[NotifyTime.PROCESSINSTANCE_TIMEOUT] = 'PROCESSINSTANCE_TIMEOUT';
NotifyTime[NotifyTime.PROCESSINSTANCE_ABORT] = 'PROCESSINSTANCE_ABORT';
NotifyTime[NotifyTime.PROCESSINSTANCE_COMPLETE] = 'PROCESSINSTANCE_COMPLETE';
NotifyTime[NotifyTime.PROCESSINSTANCE_CREATE] = 'PROCESSINSTANCE_CREATE';
NotifyTime[NotifyTime.PROCESSINSTANCE_DELETE] = 'PROCESSINSTANCE_DELETE';
NotifyTime[NotifyTime.PROCESSINSTANCE_RESUME] = 'PROCESSINSTANCE_RESUME';
NotifyTime[NotifyTime.PROCESSINSTANCE_START] = 'PROCESSINSTANCE_START';
NotifyTime[NotifyTime.PROCESSINSTANCE_SUSPEND] = 'PROCESSINSTANCE_SUSPEND';
NotifyTime[NotifyTime.PROCESSINSTANCE_TERMINATE] = 'PROCESSINSTANCE_TERMINATE';
NotifyTime[NotifyTime.ACTIVITY_ABORTED] = 'ACTIVITY_ABORTED';
NotifyTime[NotifyTime.ACTIVITY_COMPLETED] = 'ACTIVITY_COMPLETED';
NotifyTime[NotifyTime.ACTIVITY_CREATED] = 'ACTIVITY_CREATED';
NotifyTime[NotifyTime.ACTIVITY_RESUMED] = 'ACTIVITY_RESUMED';
NotifyTime[NotifyTime.ACTIVITY_STARTED] = 'ACTIVITY_STARTED';
NotifyTime[NotifyTime.ACTIVITY_STOPPED] = 'ACTIVITY_STOPPED';
NotifyTime[NotifyTime.ACTIVITY_SUSPENDED] = 'ACTIVITY_SUSPENDED';
NotifyTime[NotifyTime.ACTIVITY_TERMINATED] = 'ACTIVITY_TERMINATED';
NotifyTime[NotifyTime.ACTIVITY_RETRIEVED] = 'ACTIVITY_RETRIEVED';
NotifyTime[NotifyTime.ACTIVITY_RETRIEVED_FROM] = 'ACTIVITY_RETRIEVED_FROM';
NotifyTime[NotifyTime.ACTIVITY_ROLLBACK] = 'ACTIVITY_ROLLBACK';
NotifyTime[NotifyTime.ACTIVITY_ROLLBACK_TO] = 'ACTIVITY_ROLLBACK_TO';
NotifyTime[NotifyTime.WORKITEM_ABORTED] = 'WORKITEM_ABORTED';
NotifyTime[NotifyTime.WORKITEM_ASSIGNED] = 'WORKITEM_ASSIGNED';
NotifyTime[NotifyTime.WORKITEM_COMPLETED] = 'WORKITEM_COMPLETED';
NotifyTime[NotifyTime.WORKITEM_CREATED] = 'WORKITEM_CREATED';
NotifyTime[NotifyTime.WORKITEM_RESUMED] = 'WORKITEM_RESUMED';
NotifyTime[NotifyTime.WORKITEM_STARTED] = 'WORKITEM_STARTED';
NotifyTime[NotifyTime.WORKITEM_STOPPED] = 'WORKITEM_STOPPED';
NotifyTime[NotifyTime.WORKITEM_SUSPENDED] = 'WORKITEM_SUSPENDED';
NotifyTime[NotifyTime.WORKITEM_TERMINATED] = 'WORKITEM_TERMINATED';
NotifyTime[NotifyTime.WORKITEM_ALARM] = 'WORKITEM_ALARM';
NotifyTime[NotifyTime.WORKITEM_TIMEOUT] = 'WORKITEM_TIMEOUT';
NotifyTime[NotifyTime.UNDEFINITION] = 'UNDEFINITION';
