// @gsp-wf/wf-process-model/esm5/lib/model/extension/OtherAssignNodes.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { OtherAssignActivityNodeInfo } from "./OtherAssignActivityNodeInfo";
var OtherAssignNodes = /** @class */ (function (_super) {
    tslib_1.__extends(OtherAssignNodes, _super);
    function OtherAssignNodes(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = OtherAssignNodes.Clr_Type_ID;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    OtherAssignNodes.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.otherAssignActivityNodeInfos = new Array();
    };
    /**
     * @return {?}
     */
    OtherAssignNodes.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        if (this.otherAssignActivityNodeInfos && this.otherAssignActivityNodeInfos.length > 0) {
            /** @type {?} */
            var destArray = new Array();
            try {
                for (var _b = tslib_1.__values(this.otherAssignActivityNodeInfos), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var item = _c.value;
                    destArray.push(item.ConvertToJsonObject());
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            BpmnModelHelper.SetAttributeValue(obj, "OtherAssignActivityNodeInfos", destArray);
        }
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    OtherAssignNodes.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var e_2, _a, e_3, _b;
        _super.prototype.LoadFromJsonObject.call(this, obj);
        /** @type {?} */
        var destArrayElement = ExtensionMethod.GetAttributeValue(obj, "OtherAssignActivities");
        if (destArrayElement && destArrayElement.length > 0) {
            try {
                for (var destArrayElement_1 = tslib_1.__values(destArrayElement), destArrayElement_1_1 = destArrayElement_1.next(); !destArrayElement_1_1.done; destArrayElement_1_1 = destArrayElement_1.next()) {
                    var item = destArrayElement_1_1.value;
                    this.otherAssignActivities.push(item);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (destArrayElement_1_1 && !destArrayElement_1_1.done && (_a = destArrayElement_1.return)) _a.call(destArrayElement_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        /** @type {?} */
        var newDestArrayElement = ExtensionMethod.GetAttributeValue(obj, "OtherAssignActivityNodeInfos");
        if (newDestArrayElement && newDestArrayElement.length > 0) {
            try {
                for (var newDestArrayElement_1 = tslib_1.__values(newDestArrayElement), newDestArrayElement_1_1 = newDestArrayElement_1.next(); !newDestArrayElement_1_1.done; newDestArrayElement_1_1 = newDestArrayElement_1.next()) {
                    var item = newDestArrayElement_1_1.value;
                    /** @type {?} */
                    var otherInfo = new OtherAssignActivityNodeInfo(this.Model);
                    otherInfo.LoadFromJsonObject(item);
                    this.otherAssignActivityNodeInfos.push(otherInfo);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (newDestArrayElement_1_1 && !newDestArrayElement_1_1.done && (_b = newDestArrayElement_1.return)) _b.call(newDestArrayElement_1);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
    };
    /**
     * @return {?}
     */
    OtherAssignNodes.prototype.clone = /**
     * @return {?}
     */
    function () {
        var e_4, _a, e_5, _b;
        /** @type {?} */
        var assignScope = new OtherAssignNodes(this.Model);
        assignScope.otherAssignActivities = new Array();
        if (this.otherAssignActivities && this.otherAssignActivities.length > 0) {
            try {
                for (var _c = tslib_1.__values(this.otherAssignActivities), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var d = _d.value;
                    assignScope.otherAssignActivities.push(d);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
        assignScope.otherAssignActivityNodeInfos = new Array();
        if (this.otherAssignActivityNodeInfos && this.otherAssignActivityNodeInfos.length > 0) {
            try {
                for (var _e = tslib_1.__values(this.otherAssignActivityNodeInfos), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var d = _f.value;
                    assignScope.otherAssignActivityNodeInfos.push(d.clone());
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                }
                finally { if (e_5) throw e_5.error; }
            }
        }
        return assignScope;
    };
    /**
     * @return {?}
     */
    OtherAssignNodes.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new OtherAssignNodes(this.Model);
    };
    OtherAssignNodes.Clr_Type_ID = "wf.OtherAssignNodes";
    return OtherAssignNodes;
}(ExtensionElement));
export { OtherAssignNodes };
if (false) {
    /** @type {?} */
    OtherAssignNodes.Clr_Type_ID;
    /** @type {?} */
    OtherAssignNodes.prototype.ClrTypeID;
    /** @type {?} */
    OtherAssignNodes.prototype.otherAssignActivities;
    /** @type {?} */
    OtherAssignNodes.prototype.otherAssignActivityNodeInfos;
}
BpmnModelHelper.RegisterElementType(new OtherAssignNodes(new BpmnModel()));
