// @gsp-wf/wf-process-model/esm5/lib/model/extension/OtherAssignActivityNodeInfo.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { Condition } from "./condition-expression/Condition";
var OtherAssignActivityNodeInfo = /** @class */ (function (_super) {
    tslib_1.__extends(OtherAssignActivityNodeInfo, _super);
    function OtherAssignActivityNodeInfo(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = OtherAssignActivityNodeInfo.Clr_Type_ID;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    OtherAssignActivityNodeInfo.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    OtherAssignActivityNodeInfo.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "ActivityDefinitionId", this.activityDefinitionId);
        BpmnModelHelper.SetAttributeValue(obj, "Required", this.required);
        if (this.condition) {
            BpmnModelHelper.SetAttributeValue(obj, "Condition", this.condition.ConvertToJsonObject());
        }
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    OtherAssignActivityNodeInfo.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.activityDefinitionId = ExtensionMethod.GetAttributeValue4String(obj, "ActivityDefinitionId");
        this.required = ExtensionMethod.GetAttributeValue4Boolean(obj, "Required");
        /** @type {?} */
        var conditionElement = ExtensionMethod.GetAttributeValue(obj, 'Condition');
        if (conditionElement) {
            /** @type {?} */
            var condition = new Condition(this.Model);
            condition.LoadFromJsonObject(conditionElement);
            this.condition = condition;
        }
    };
    /**
     * @return {?}
     */
    OtherAssignActivityNodeInfo.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var otherAssignActivityNodeInfo = new OtherAssignActivityNodeInfo(this.Model);
        otherAssignActivityNodeInfo.activityDefinitionId = this.activityDefinitionId;
        if (this.condition) {
            otherAssignActivityNodeInfo.condition = this.condition.clone();
        }
        otherAssignActivityNodeInfo.required = this.required;
        return otherAssignActivityNodeInfo;
    };
    /**
     * @return {?}
     */
    OtherAssignActivityNodeInfo.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new OtherAssignActivityNodeInfo(this.Model);
    };
    OtherAssignActivityNodeInfo.Clr_Type_ID = "wf.OtherAssignActivityNodeInfo";
    return OtherAssignActivityNodeInfo;
}(ExtensionElement));
export { OtherAssignActivityNodeInfo };
if (false) {
    /** @type {?} */
    OtherAssignActivityNodeInfo.Clr_Type_ID;
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.ClrTypeID;
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.activityDefinitionId;
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.condition;
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.required;
}
BpmnModelHelper.RegisterElementType(new OtherAssignActivityNodeInfo(new BpmnModel()));
