// @gsp-wf/wf-process-model/esm5/lib/model/extension/ProcessHeader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionElement } from "@edp-pmf/bpmn-model";
import { StartMode } from "./StartMode";
var ProcessHeader = /** @class */ (function (_super) {
    tslib_1.__extends(ProcessHeader, _super);
    function ProcessHeader(model, bizDefKey, bizObjId, bizDefName, startMode) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "wf.ProcessHeader";
        _this.BizDefKey = bizDefKey || '';
        _this.BizObjId = bizObjId || '';
        _this.BizDefName = bizDefName || '';
        _this.allowPreAssign = false;
        _this.enablePriority = false;
        _this.enableWholeOrderRetrieve = false;
        _this.startMode = startMode || StartMode.DIRECT;
        return _this;
    }
    /**
     * @return {?}
     */
    ProcessHeader.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new ProcessHeader(this.Model);
    };
    /**
     * @protected
     * @return {?}
     */
    ProcessHeader.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.BizDefKey = null;
        this.BizObjId = null;
        this.BizDefName = null;
        this.allowPreAssign = false;
        this.enablePriority = false;
        this.enableWholeOrderRetrieve = false;
    };
    /**
     * @return {?}
     */
    ProcessHeader.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "BizDefKey", this.BizDefKey);
        BpmnModelHelper.SetAttributeValue(obj, "BizObjId", this.BizObjId);
        BpmnModelHelper.SetAttributeValue(obj, "BizDefName", this.BizDefName);
        BpmnModelHelper.SetAttributeValue(obj, "AllowPreAssign", this.allowPreAssign);
        BpmnModelHelper.SetAttributeValue(obj, "EnablePriority", this.enablePriority);
        BpmnModelHelper.SetAttributeValue(obj, "EnableWholeOrderRetrieve", this.enableWholeOrderRetrieve);
        BpmnModelHelper.SetAttributeValue(obj, "StartMode", this.startMode);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ProcessHeader.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.BizDefKey = obj.BizDefKey;
        this.BizObjId = obj.BizObjId;
        this.BizDefName = obj.BizDefName;
        this.allowPreAssign = obj.AllowPreAssign;
        this.enablePriority = obj.EnablePriority;
        this.enableWholeOrderRetrieve = obj.EnableWholeOrderRetrieve;
        if (obj.StartMode) {
            this.startMode = obj.StartMode;
        }
        else if (obj.AccessLevel) {
            switch (obj.AccessLevel) {
                case 'PUBLIC':
                    this.startMode = StartMode.DIRECT;
                    break;
                case 'PROTECTED':
                    this.startMode = StartMode.EXTERNAL_PROCESS;
                    break;
                case 'PRIVATE':
                    this.startMode = StartMode.SUB_PROCESS;
                    break;
            }
        }
    };
    return ProcessHeader;
}(ExtensionElement));
export { ProcessHeader };
if (false) {
    /** @type {?} */
    ProcessHeader.prototype.ClrTypeID;
    /** @type {?} */
    ProcessHeader.prototype.BizDefKey;
    /** @type {?} */
    ProcessHeader.prototype.BizDefName;
    /** @type {?} */
    ProcessHeader.prototype.BizObjId;
    /** @type {?} */
    ProcessHeader.prototype.allowPreAssign;
    /** @type {?} */
    ProcessHeader.prototype.enablePriority;
    /** @type {?} */
    ProcessHeader.prototype.enableWholeOrderRetrieve;
    /** @type {?} */
    ProcessHeader.prototype.startMode;
}
BpmnModelHelper.RegisterElementType(new ProcessHeader(new BpmnModel()));
