// @gsp-wf/wf-process-model/esm5/lib/model/extension/WFExtensionElement.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelHelper, ExtensionElement } from "@edp-pmf/bpmn-model";
/**
 * @abstract
 */
var /**
 * @abstract
 */
WFExtensionElement = /** @class */ (function (_super) {
    tslib_1.__extends(WFExtensionElement, _super);
    function WFExtensionElement(model, id, name) {
        var _this = _super.call(this, model) || this;
        _this.ExtendedAttributes = {};
        _this.Id = id || '';
        _this.Name = name || '';
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    WFExtensionElement.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.ExtendedAttributes = {};
    };
    /**
     * @return {?}
     */
    WFExtensionElement.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "Id", this.Id);
        BpmnModelHelper.SetAttributeValue(obj, "Code", this.Code);
        BpmnModelHelper.SetAttributeValue(obj, "Name", this.Name);
        BpmnModelHelper.SetAttributeValue(obj, "Description", this.Description);
        /** @type {?} */
        var extAttributeElements = [];
        for (var key in this.ExtendedAttributes) {
            /** @type {?} */
            var obj_1 = {};
            BpmnModelHelper.SetAttributeValue(obj_1, "Name", key);
            BpmnModelHelper.SetAttributeValue(obj_1, "Value", this.ExtendedAttributes[key]);
            extAttributeElements.push(obj_1);
        }
        BpmnModelHelper.SetAttributeValue(obj, "ExtendedAttributes", extAttributeElements);
        if (this.GetMexExtensionElements4ConvertTo().length > 0) {
            /** @type {?} */
            var mexJsonArray_1 = [];
            this.GetMexExtensionElements4ConvertTo().forEach((/**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                mexJsonArray_1.push(value.ConvertToJsonObject());
            }));
            BpmnModelHelper.SetAttributeValue(obj, "MexExtensionElements", mexJsonArray_1);
        }
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    WFExtensionElement.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Id = obj.Id;
        this.Code = obj.Code;
        this.Name = obj.Name;
        this.Description = obj.Description;
        /** @type {?} */
        var extAttributeElements = (/** @type {?} */ (obj.ExtendedAttributes));
        if (extAttributeElements && extAttributeElements.length > 0) {
            extAttributeElements.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                _this.ExtendedAttributes[item.Name] = item.Value;
            }));
        }
        /** @type {?} */
        var mexJsonArray = (/** @type {?} */ (obj.MexExtensionElements));
        if (mexJsonArray && mexJsonArray.length > 0) {
            /** @type {?} */
            var list_1 = new Array();
            mexJsonArray.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                /** @type {?} */
                var mexExtElementObj = (/** @type {?} */ (BpmnModelHelper.GetElementType(item.ClrTypeID)));
                if (mexExtElementObj) {
                    mexExtElementObj.LoadFromJsonObject(item);
                    list_1.push(mexExtElementObj);
                }
            }));
            this.SetMexExtensionElements4LoadFrom(list_1);
        }
    };
    /**
     * @protected
     * @return {?}
     */
    WFExtensionElement.prototype.GetMexExtensionElements4ConvertTo = /**
     * @protected
     * @return {?}
     */
    function () {
        return [];
    };
    /**
     * @protected
     * @param {?} list
     * @return {?}
     */
    WFExtensionElement.prototype.SetMexExtensionElements4LoadFrom = /**
     * @protected
     * @param {?} list
     * @return {?}
     */
    function (list) {
    };
    /**
     * @return {?}
     */
    WFExtensionElement.prototype.GetExtendedAttributes = /**
     * @return {?}
     */
    function () {
        return this.ExtendedAttributes;
    };
    return WFExtensionElement;
}(ExtensionElement));
/**
 * @abstract
 */
export { WFExtensionElement };
if (false) {
    /** @type {?} */
    WFExtensionElement.prototype.Id;
    /** @type {?} */
    WFExtensionElement.prototype.Code;
    /** @type {?} */
    WFExtensionElement.prototype.Name;
    /** @type {?} */
    WFExtensionElement.prototype.Description;
    /** @type {?} */
    WFExtensionElement.prototype.ExtendedAttributes;
}
