// @edp-bif/runtime-api/esm5/lib/entity/tracking/BillChildInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/BillChildInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { BillItemInfo } from "./BillItemInfo";
var BillChildInfo = /** @class */ (function (_super) {
    tslib_1.__extends(BillChildInfo, _super);
    function BillChildInfo() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.billItemInfos = [];
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    BillChildInfo.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.itemCode = null;
        this.billItemInfos = [];
    };
    /**
     * @return {?}
     */
    BillChildInfo.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "itemCode", this.itemCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billItemInfos", this.billItemInfos);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BillChildInfo.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.itemCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "itemCode");
        this.billItemInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "billItemInfos", BillItemInfo);
    };
    return BillChildInfo;
}(JsonConvertableImpl));
export { BillChildInfo };
if (false) {
    /** @type {?} */
    BillChildInfo.prototype.itemCode;
    /** @type {?} */
    BillChildInfo.prototype.billItemInfos;
}
