// @edp-bif/runtime-api/esm5/lib/entity/tracking/BillDetailInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/BillDetailInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
var BillDetailInfo = /** @class */ (function (_super) {
    tslib_1.__extends(BillDetailInfo, _super);
    function BillDetailInfo() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.detailIds = [];
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    BillDetailInfo.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.detailCode = null;
        this.detailIds = [];
    };
    /**
     * @return {?}
     */
    BillDetailInfo.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "detailCode", this.detailCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "detailIds", BpmnModelJsonHelper.StringListToJson(this.detailIds));
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BillDetailInfo.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.detailCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "detailCode");
        this.detailIds = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "detailIds"));
    };
    return BillDetailInfo;
}(JsonConvertableImpl));
export { BillDetailInfo };
if (false) {
    /** @type {?} */
    BillDetailInfo.prototype.detailCode;
    /** @type {?} */
    BillDetailInfo.prototype.detailIds;
}
