// @edp-bif/runtime-api/esm5/lib/entity/tracking/BillItemInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/BillItemInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { BillDetailInfo } from "./BillDetailInfo";
var BillItemInfo = /** @class */ (function (_super) {
    tslib_1.__extends(BillItemInfo, _super);
    function BillItemInfo() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.billDetailInfos = [];
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    BillItemInfo.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.itemId = null;
        this.billDetailInfos = [];
    };
    /**
     * @return {?}
     */
    BillItemInfo.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "itemId", this.itemId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billDetailInfos", this.billDetailInfos);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BillItemInfo.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.itemId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "itemId");
        this.billDetailInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "billDetailInfos", BillDetailInfo);
    };
    return BillItemInfo;
}(JsonConvertableImpl));
export { BillItemInfo };
if (false) {
    /** @type {?} */
    BillItemInfo.prototype.itemId;
    /** @type {?} */
    BillItemInfo.prototype.billDetailInfos;
}
