// @edp-bif/runtime-api/esm5/lib/entity/tracking/BillInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/BillInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { JointQueryMode } from '@edp-bif/common-api';
import { BillDataInfo } from './BillDataInfo';
var BillInfo = /** @class */ (function (_super) {
    tslib_1.__extends(BillInfo, _super);
    function BillInfo() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.billDataInfos = [];
        _this.jointQueryMode = JointQueryMode.DEFAULT;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    BillInfo.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.billTrackingSettings = null;
        this.schemaId = null;
        this.suCode = null;
        this.documentTypeId = null;
        this.bizKind = null;
        this.bizType = null;
        this.billDataInfos = [];
        this.jointQueryMode = JointQueryMode.DEFAULT;
        this.jointQueryParameters = null;
        this.trackingSourceType = null;
    };
    /**
     * @return {?}
     */
    BillInfo.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billTrackingSettings", this.billTrackingSettings);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaId", this.schemaId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "documentTypeId", this.documentTypeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizKind", this.bizKind);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizType", this.bizType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billDataInfos", this.billDataInfos);
        BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryMode", this.jointQueryMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryParameters", this.jointQueryParameters);
        BpmnModelJsonHelper.SetAttributeValue(obj, "trackingSourceType", this.trackingSourceType);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BillInfo.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.billTrackingSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billTrackingSettings");
        this.schemaId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaId");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.documentTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "documentTypeId");
        this.bizKind = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizKind");
        this.bizType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizType");
        this.billDataInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "billDataInfos", BillDataInfo);
        this.jointQueryMode = JointQueryMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryMode")];
        this.jointQueryParameters = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryParameters");
        this.trackingSourceType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "trackingSourceType");
    };
    return BillInfo;
}(JsonConvertableImpl));
export { BillInfo };
if (false) {
    /** @type {?} */
    BillInfo.prototype.billTrackingSettings;
    /** @type {?} */
    BillInfo.prototype.schemaId;
    /** @type {?} */
    BillInfo.prototype.suCode;
    /** @type {?} */
    BillInfo.prototype.documentTypeId;
    /** @type {?} */
    BillInfo.prototype.bizKind;
    /** @type {?} */
    BillInfo.prototype.bizType;
    /** @type {?} */
    BillInfo.prototype.billDataInfos;
    /** @type {?} */
    BillInfo.prototype.jointQueryMode;
    /** @type {?} */
    BillInfo.prototype.jointQueryParameters;
    /** @type {?} */
    BillInfo.prototype.trackingSourceType;
}
