// @edp-bif/runtime-api/esm5/lib/entity/tracking/DecimalTypeFormatInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/DecimalTypeFormatInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { ScaleMode } from "./ScaleMode";
var DecimalTypeFormatInfo = /** @class */ (function (_super) {
    tslib_1.__extends(DecimalTypeFormatInfo, _super);
    function DecimalTypeFormatInfo() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.scaleMode = ScaleMode.DEFAULT;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    DecimalTypeFormatInfo.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.scale = 0;
        this.scaleMode = ScaleMode.DEFAULT;
        this.fieldPath = null;
        this.version = 0;
    };
    /**
     * @return {?}
     */
    DecimalTypeFormatInfo.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "scale", this.scale);
        BpmnModelJsonHelper.SetAttributeValue(obj, "scale", this.scale);
        BpmnModelJsonHelper.SetAttributeValue(obj, "scaleMode", this.scaleMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "fieldPath", this.fieldPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "version", this.version);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    DecimalTypeFormatInfo.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.scale = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "scale");
        /** @type {?} */
        var scaleMode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "scaleMode");
        this.scaleMode = scaleMode == null ? ScaleMode.DEFAULT : ScaleMode[scaleMode];
        this.fieldPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "fieldPath");
        this.version = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "version");
    };
    return DecimalTypeFormatInfo;
}(JsonConvertableImpl));
export { DecimalTypeFormatInfo };
if (false) {
    /** @type {?} */
    DecimalTypeFormatInfo.prototype.scale;
    /** @type {?} */
    DecimalTypeFormatInfo.prototype.scaleMode;
    /** @type {?} */
    DecimalTypeFormatInfo.prototype.fieldPath;
    /** @type {?} */
    DecimalTypeFormatInfo.prototype.version;
}
