// @edp-bif/runtime-api/esm5/lib/entity/tracking/FieldsWithOrder.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/FieldsWithOrder.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { TypeToFormat } from "./typeToFormat";
var FieldsWithOrder = /** @class */ (function (_super) {
    tslib_1.__extends(FieldsWithOrder, _super);
    function FieldsWithOrder() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * 数据类型
         */
        _this.typeToFormat = TypeToFormat.DEFAULT;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    FieldsWithOrder.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.order = 0;
        this.fieldName = null;
        this.fieldPath = null;
        this.name = null;
        this.customName = null;
        this.colI18nValues = null;
        this.typeToFormat = TypeToFormat.DEFAULT;
        this.formatInfo = null;
        this.version = 0;
    };
    /**
     * @return {?}
     */
    FieldsWithOrder.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "order", this.order);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "fieldName", this.fieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "fieldPath", this.fieldPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "customName", this.customName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "colI18nValues", this.colI18nValues);
        BpmnModelJsonHelper.SetAttributeValue(obj, "typeToFormat", this.typeToFormat);
        BpmnModelJsonHelper.SetAttributeValue(obj, "formatInfo", this.formatInfo);
        BpmnModelJsonHelper.SetAttributeValue(obj, "version", this.version);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FieldsWithOrder.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.order = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "order");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.fieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "fieldName");
        this.fieldPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "fieldPath");
        this.customName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "customName");
        this.colI18nValues = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "colI18nValues");
        this.typeToFormat = TypeToFormat[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "typeToFormat")];
        this.formatInfo = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formatInfo");
        this.version = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "version");
    };
    return FieldsWithOrder;
}(JsonConvertableImpl));
export { FieldsWithOrder };
if (false) {
    /** @type {?} */
    FieldsWithOrder.prototype.order;
    /**
     * 字段名
     * @type {?}
     */
    FieldsWithOrder.prototype.fieldName;
    /**
     * 字段名称
     * @type {?}
     */
    FieldsWithOrder.prototype.name;
    /**
     * 字段路径
     * @type {?}
     */
    FieldsWithOrder.prototype.fieldPath;
    /**
     * 自定义名称
     * @type {?}
     */
    FieldsWithOrder.prototype.customName;
    /**
     * 多语信息
     * @type {?}
     */
    FieldsWithOrder.prototype.colI18nValues;
    /**
     * 数据类型
     * @type {?}
     */
    FieldsWithOrder.prototype.typeToFormat;
    /**
     * 格式化
     * @type {?}
     */
    FieldsWithOrder.prototype.formatInfo;
    /**
     * 版本信息
     * @type {?}
     */
    FieldsWithOrder.prototype.version;
}
