// @edp-bif/runtime-api/esm5/lib/entity/tracking/GetBillsInfoArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/GetBillsInfoArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { BillChildInfo } from './BillChildInfo';
var GetBillsInfoArgs = /** @class */ (function (_super) {
    tslib_1.__extends(GetBillsInfoArgs, _super);
    function GetBillsInfoArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.billChildInfos = [];
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    GetBillsInfoArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.documentTypeId = null;
        this.bizKind = null;
        this.bizType = null;
        this.billDataId = null;
        this.billChildInfos = [];
        this.curYear = null;
    };
    /**
     * @return {?}
     */
    GetBillsInfoArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "documentTypeId", this.documentTypeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizKind", this.bizKind);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizType", this.bizType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billDataId", this.billDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billChildInfos", this.billChildInfos);
        BpmnModelJsonHelper.SetAttributeValue(obj, "curYear", this.curYear);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    GetBillsInfoArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.documentTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "documentTypeId");
        this.bizKind = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizKind");
        this.bizType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizType");
        this.billDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billDataId");
        this.billChildInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "billChildInfos", BillChildInfo);
        this.curYear = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "curYear");
    };
    return GetBillsInfoArgs;
}(JsonConvertableImpl));
export { GetBillsInfoArgs };
if (false) {
    /** @type {?} */
    GetBillsInfoArgs.prototype.documentTypeId;
    /** @type {?} */
    GetBillsInfoArgs.prototype.bizKind;
    /** @type {?} */
    GetBillsInfoArgs.prototype.bizType;
    /** @type {?} */
    GetBillsInfoArgs.prototype.billDataId;
    /** @type {?} */
    GetBillsInfoArgs.prototype.billChildInfos;
    /** @type {?} */
    GetBillsInfoArgs.prototype.curYear;
}
