// @edp-bif/runtime-api/esm5/lib/entity/tracking/JointQueryParametersContent.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/JointQueryParametersContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { JointQueryParameterMode } from "./JointQueryParameterMode";
var JointQueryParametersContent = /** @class */ (function (_super) {
    tslib_1.__extends(JointQueryParametersContent, _super);
    function JointQueryParametersContent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.jointQueryParameterMode = JointQueryParameterMode.FIELD;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    JointQueryParametersContent.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.jointQueryParameterMode = JointQueryParameterMode.FIELD;
        this.content = null;
    };
    /**
     * @return {?}
     */
    JointQueryParametersContent.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryParameterMode", this.jointQueryParameterMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    JointQueryParametersContent.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.jointQueryParameterMode = JointQueryParameterMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryParameterMode")];
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
    };
    return JointQueryParametersContent;
}(JsonConvertableImpl));
export { JointQueryParametersContent };
if (false) {
    /** @type {?} */
    JointQueryParametersContent.prototype.jointQueryParameterMode;
    /** @type {?} */
    JointQueryParametersContent.prototype.content;
}
