// @edp-bif/runtime-api/esm5/lib/entity/tracking/SimpleValueMappingInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/SimpleValueMappingInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
var SimpleValueMappingInfo = /** @class */ (function (_super) {
    tslib_1.__extends(SimpleValueMappingInfo, _super);
    function SimpleValueMappingInfo() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    SimpleValueMappingInfo.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.displayValue = null;
        this.actualValue = null;
    };
    /**
     * @return {?}
     */
    SimpleValueMappingInfo.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "displayValue", this.displayValue);
        BpmnModelJsonHelper.SetAttributeValue(obj, "actualValue", this.actualValue);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    SimpleValueMappingInfo.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.displayValue = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "displayValue");
        this.actualValue = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "actualValue");
    };
    return SimpleValueMappingInfo;
}(JsonConvertableImpl));
export { SimpleValueMappingInfo };
if (false) {
    /** @type {?} */
    SimpleValueMappingInfo.prototype.displayValue;
    /** @type {?} */
    SimpleValueMappingInfo.prototype.actualValue;
}
