// @edp-bif/runtime-api/esm5/lib/entity/tracking/StringTypeFormatInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/StringTypeFormatInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { SimpleValueMappingInfo } from "./SimpleValueMappingInfo";
import { StringDisplayMode } from "./StringDisplayMode";
var StringTypeFormatInfo = /** @class */ (function (_super) {
    tslib_1.__extends(StringTypeFormatInfo, _super);
    function StringTypeFormatInfo() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.stringDisplayMode = StringDisplayMode.SIMPLEVALUEMAPPING;
        _this.simpleValueMappingInfos = [];
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    StringTypeFormatInfo.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.stringDisplayMode = StringDisplayMode.SIMPLEVALUEMAPPING;
        this.simpleValueMappingInfos = [];
    };
    /**
     * @return {?}
     */
    StringTypeFormatInfo.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "stringDisplayMode", this.stringDisplayMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "simpleValueMappingInfos", this.simpleValueMappingInfos);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    StringTypeFormatInfo.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.stringDisplayMode = StringDisplayMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "stringDisplayMode")];
        this.simpleValueMappingInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "simpleValueMappingInfos", SimpleValueMappingInfo);
    };
    return StringTypeFormatInfo;
}(JsonConvertableImpl));
export { StringTypeFormatInfo };
if (false) {
    /** @type {?} */
    StringTypeFormatInfo.prototype.stringDisplayMode;
    /** @type {?} */
    StringTypeFormatInfo.prototype.simpleValueMappingInfos;
}
