// @edp-bif/runtime-api/esm5/lib/entity/tracking/SummaryFieldsInfo.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/SummaryFieldsInfo.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
import { FieldsWithOrder } from "./FieldsWithOrder";
var SummaryFieldsInfo = /** @class */ (function (_super) {
    tslib_1.__extends(SummaryFieldsInfo, _super);
    function SummaryFieldsInfo() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.fields = [];
        _this.childFieldInfos = [];
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    SummaryFieldsInfo.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.code = null;
        this.fields = [];
        this.childFieldInfos = [];
    };
    /**
     * @return {?}
     */
    SummaryFieldsInfo.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "fields", this.fields);
        BpmnModelJsonHelper.SetAttributeValue(obj, "childFieldInfos", this.childFieldInfos);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    SummaryFieldsInfo.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.fields = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "fields", FieldsWithOrder);
        this.childFieldInfos = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "childFieldInfos", SummaryFieldsInfo);
    };
    return SummaryFieldsInfo;
}(JsonConvertableImpl));
export { SummaryFieldsInfo };
if (false) {
    /** @type {?} */
    SummaryFieldsInfo.prototype.code;
    /** @type {?} */
    SummaryFieldsInfo.prototype.fields;
    /** @type {?} */
    SummaryFieldsInfo.prototype.childFieldInfos;
}
