// @edp-bif/runtime-api/esm5/lib/entity/tracking/TrackingMappingAssociation.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/tracking/TrackingMappingAssociation.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
var TrackingMappingAssociation = /** @class */ (function (_super) {
    tslib_1.__extends(TrackingMappingAssociation, _super);
    function TrackingMappingAssociation() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    TrackingMappingAssociation.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.upBillTableCodePath = null;
        this.upBillDataIdPath = null;
        this.downBillTableCodePath = null;
        this.downBillDataIdPath = null;
    };
    /**
     * @return {?}
     */
    TrackingMappingAssociation.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTableCodePath", this.upBillTableCodePath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillDataIdPath", this.upBillDataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTableCodePath", this.downBillTableCodePath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillDataIdPath", this.downBillDataIdPath);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    TrackingMappingAssociation.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.upBillTableCodePath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTableCodePath");
        this.upBillDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillDataIdPath");
        this.downBillTableCodePath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTableCodePath");
        this.downBillDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillDataIdPath");
    };
    return TrackingMappingAssociation;
}(JsonConvertableImpl));
export { TrackingMappingAssociation };
if (false) {
    /** @type {?} */
    TrackingMappingAssociation.prototype.upBillTableCodePath;
    /** @type {?} */
    TrackingMappingAssociation.prototype.upBillDataIdPath;
    /** @type {?} */
    TrackingMappingAssociation.prototype.downBillTableCodePath;
    /** @type {?} */
    TrackingMappingAssociation.prototype.downBillDataIdPath;
}
