// @gsp-sys/rac-common/esm5/lib/common/entity/resultcontext.entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ResultContextEntity = /** @class */ (function () {
    function ResultContextEntity() {
        /**
         * 组织岗上的数据权限自动继承业务组织权限
         */
        this.inheritPermission = false;
        /**
         * 维度1的值
         */
        this.assid1 = '';
        /**
         * 维度2的值
         */
        this.assid2 = '';
        /**
         * 变量值
         */
        this.auVarValues = [];
    }
    return ResultContextEntity;
}());
export { ResultContextEntity };
if (false) {
    /** @type {?} */
    ResultContextEntity.prototype.ownerId;
    /** @type {?} */
    ResultContextEntity.prototype.ownerCode;
    /** @type {?} */
    ResultContextEntity.prototype.ownerName;
    /**
     * 权限所有者类型 岗位1，用户2
     * @type {?}
     */
    ResultContextEntity.prototype.ownerType;
    /**
     * 岗位类型 组织岗0，通用岗1，用户2
     * @type {?}
     */
    ResultContextEntity.prototype.posType;
    /** @type {?} */
    ResultContextEntity.prototype.packageId;
    /**
     * 组织岗上的数据权限自动继承业务组织权限
     * @type {?}
     */
    ResultContextEntity.prototype.inheritPermission;
    /** @type {?} */
    ResultContextEntity.prototype.yearObj;
    /** @type {?} */
    ResultContextEntity.prototype.year;
    /**
     * 维度1的值
     * @type {?}
     */
    ResultContextEntity.prototype.assid1;
    /**
     * 维度2的值
     * @type {?}
     */
    ResultContextEntity.prototype.assid2;
    /**
     * 变量值
     * @type {?}
     */
    ResultContextEntity.prototype.auVarValues;
    /** @type {?} */
    ResultContextEntity.prototype.authfield;
    /** @type {?} */
    ResultContextEntity.prototype.authObjId;
}
/**
 * 权限字段用到的变量键值对
 */
var /**
 * 权限字段用到的变量键值对
 */
AuthfieldVarEntity = /** @class */ (function () {
    function AuthfieldVarEntity() {
    }
    return AuthfieldVarEntity;
}());
/**
 * 权限字段用到的变量键值对
 */
export { AuthfieldVarEntity };
if (false) {
    /** @type {?} */
    AuthfieldVarEntity.prototype.key;
    /** @type {?} */
    AuthfieldVarEntity.prototype.value;
}
