// @edp-bif/runtime-api/esm2015/lib/entity/form-rest/FormBillCreationProcessExecutorArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/form-rest/FormBillCreationProcessExecutorArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from "@edp-aif/common-api";
import { FormCreationRuleExecutorArgs } from "@edp-aif/runtime-api";
import { ManualCreationMode } from "./ManualCreationMode";
export class FormBillCreationProcessExecutorArgs extends FormCreationRuleExecutorArgs {
    constructor() {
        super(...arguments);
        /**
         * sourceVoEntityDatas和sourceVoEntityDataIds设置其一即可，优先使用sourceVoEntityDatas，若为空则根据sourceVoEntityDataIds查询
         */
        this.sourceVoEntityDataIds = [];
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.billCreationProcessId = null;
        this.creationRuleInstanceId = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "billCreationProcessId", this.billCreationProcessId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "manualCreationMode", this.manualCreationMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntityDataIds", BpmnModelJsonHelper.StringListToJson(this.sourceVoEntityDataIds));
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.billCreationProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billCreationProcessId");
        this.creationRuleInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
        this.manualCreationMode = ManualCreationMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId")];
        this.sourceVoEntityDataIds = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValue(obj, "sourceVoEntityDataIds"));
    }
}
if (false) {
    /** @type {?} */
    FormBillCreationProcessExecutorArgs.prototype.billCreationProcessId;
    /** @type {?} */
    FormBillCreationProcessExecutorArgs.prototype.creationRuleInstanceId;
    /** @type {?} */
    FormBillCreationProcessExecutorArgs.prototype.manualCreationMode;
    /**
     * sourceVoEntityDatas和sourceVoEntityDataIds设置其一即可，优先使用sourceVoEntityDatas，若为空则根据sourceVoEntityDataIds查询
     * @type {?}
     */
    FormBillCreationProcessExecutorArgs.prototype.sourceVoEntityDataIds;
}
