// @edp-bif/runtime-api/esm2015/lib/entity/form-rest/FormGetBillCreationProcessEntityExArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/form-rest/FormGetBillCreationProcessEntityExArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from "@edp-aif/common-api";
import { ManualCreationMode } from './ManualCreationMode';
export class FormGetBillCreationProcessEntityExArgs extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.upBillVoId = null;
        this.downBillVoId = null;
        this.manualCreationMode = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoId", this.upBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoId", this.downBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "manualCreationMode", this.manualCreationMode);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.upBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoId");
        this.downBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoId");
        this.manualCreationMode = ManualCreationMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "manualCreationMode")];
    }
}
if (false) {
    /** @type {?} */
    FormGetBillCreationProcessEntityExArgs.prototype.upBillVoId;
    /** @type {?} */
    FormGetBillCreationProcessEntityExArgs.prototype.downBillVoId;
    /** @type {?} */
    FormGetBillCreationProcessEntityExArgs.prototype.manualCreationMode;
}
