// @edp-bif/runtime-api/esm2015/lib/entity/form-rest/GeneralTextIdWithDataId.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/form-rest/GeneralTextIdWithDataId.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from "@edp-aif/common-api";
export class GeneralTextIdWithDataId extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.generalTextId = null;
        this.dataId = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "generalTextId", this.generalTextId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dataId", this.dataId);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.generalTextId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "generalTextId");
        this.dataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dataId");
    }
}
if (false) {
    /** @type {?} */
    GeneralTextIdWithDataId.prototype.generalTextId;
    /** @type {?} */
    GeneralTextIdWithDataId.prototype.dataId;
}
