// @gspwidget/widget-devkit/esm2015/lib/widget-title-bar/widget-title-bar-select/widget-title-bar-select.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { FormControl } from '@angular/forms';
/**
 * @record
 * @template
 */
export function FilterOption() { }
if (false) {
    /** @type {?} */
    FilterOption.prototype.value;
    /** @type {?} */
    FilterOption.prototype.name;
}
export class WidgetTitleBarSelectComponent {
    constructor() {
        this.filterOptions = [];
        this.onFilterChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.filter = new FormControl(this.defaultFilterValue ? this.defaultFilterValue : this.filterOptions[0].value);
        this.filter.valueChanges.subscribe((/**
         * @param {?} newValue
         * @return {?}
         */
        newValue => {
            this.onFilterChange.emit(newValue);
        }));
    }
}
WidgetTitleBarSelectComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-widget-title-bar-select',
                template: "\n<select\n  [formControl]=\"filter\"\n  class=\"widget-select form-control\" >\n  <option *ngFor=\"let option of filterOptions\"\n    [value]=\"option.value\">{{option.name}}</option>\n</select>",
                styles: [".widget-select{font-size:12px;padding:0 5px;height:25px}"]
            }] }
];
/** @nocollapse */
WidgetTitleBarSelectComponent.ctorParameters = () => [];
WidgetTitleBarSelectComponent.propDecorators = {
    defaultFilterValue: [{ type: Input }],
    filterOptions: [{ type: Input }],
    onFilterChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    WidgetTitleBarSelectComponent.prototype.filter;
    /**
     * 若不设置，默认使用第一项 option
     * @type {?}
     */
    WidgetTitleBarSelectComponent.prototype.defaultFilterValue;
    /** @type {?} */
    WidgetTitleBarSelectComponent.prototype.filterOptions;
    /** @type {?} */
    WidgetTitleBarSelectComponent.prototype.onFilterChange;
}
