// @gsp-cmp/ubcc-param-data-ui/esm2015/lib/controls/dimension-control/dimension-dropdownlist-control/dimension-dropdownlist-control.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/controls/dimension-control/dimension-dropdownlist-control/dimension-dropdownlist-control.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { UbccEventBusService } from '../../../service/event-bus-service';
export class DimensionDropdownlistControlComponent {
    /**
     * @param {?} ebsService
     */
    constructor(ebsService) {
        this.ebsService = ebsService;
        this.valueChangeEmit = new EventEmitter();
        this.enumerationItems = [];
        this.isMultiSelect = false;
        this.isEnumInfoLoad = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.dimensionItem = this.dimensionItem || {};
        if (!this.isEnumInfoLoad) {
            this.enumConfigInit();
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (!this.isEnumInfoLoad) {
            this.enumConfigInit();
        }
    }
    /**
     * @return {?}
     */
    enumConfigInit() {
        this.getEnumItems();
    }
    /**
     * @return {?}
     */
    getEnumItems() {
        if (!this.dimensionItem || !this.dimensionItem.dataType) {
            return;
        }
        /** @type {?} */
        let enumType = (/** @type {?} */ (this.dimensionItem.dataType));
        this.get(enumType);
    }
    /**
     * @param {?} enumType
     * @return {?}
     */
    get(enumType) {
        if (!enumType || !enumType.enumerationItems) {
            return;
        }
        enumType.enumerationItems.map((/**
         * @param {?} x
         * @return {?}
         */
        x => this.enumerationItems.push(x)));
        this.isEnumInfoLoad = true;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    valueChangeHandler($event) {
        this.dimensionItem.value = $event.value;
        this.ebsService.ebs.emit(this.dimensionItem);
    }
}
DimensionDropdownlistControlComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-dimension-dropdownlist-control',
                template: "<div class=\"farris-form-controls-inline farris-group-wrap col-12\">\n    <div class=\"form-group farris-form-group\">\n      <label class=\"col-form-label col-4\">\n        <span class=\"farris-label-text\" style=\"float:right\">{{dimensionItem.name}}</span>\n      </label>\n      <div class=\"farris-input-wrap col-6\">\n        <farris-combo-list [idField]=\"'code'\" [textField]=\"'description'\" [data]=\"enumerationItems\" [multiSelect]=\"isMultiSelect\"\n          (valueChange)=\"valueChangeHandler($event)\">\n        </farris-combo-list>\n      </div>\n    </div>\n  </div>",
                styles: [""]
            }] }
];
/** @nocollapse */
DimensionDropdownlistControlComponent.ctorParameters = () => [
    { type: UbccEventBusService }
];
DimensionDropdownlistControlComponent.propDecorators = {
    dimensionItem: [{ type: Input }],
    valueChangeEmit: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DimensionDropdownlistControlComponent.prototype.dimensionItem;
    /** @type {?} */
    DimensionDropdownlistControlComponent.prototype.valueChangeEmit;
    /** @type {?} */
    DimensionDropdownlistControlComponent.prototype.enumerationItems;
    /** @type {?} */
    DimensionDropdownlistControlComponent.prototype.isMultiSelect;
    /** @type {?} */
    DimensionDropdownlistControlComponent.prototype.isEnumInfoLoad;
    /**
     * @type {?}
     * @private
     */
    DimensionDropdownlistControlComponent.prototype.ebsService;
}
