// @gsp-wf/wf-process-editor/esm2015/lib/views/execute-forms/add-execute-form/form-format.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { map } from 'rxjs/operators';
export class FormFormatService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @param {?} url
     * @param {?=} params
     * @return {?}
     */
    getData(url, params) {
        /** @type {?} */
        const customData = params.customData;
        /** @type {?} */
        const index = params.pageIndex || 1;
        /** @type {?} */
        const size = params.pageSize || 20;
        /** @type {?} */
        const searchValue = JSON.parse(params.searchValue);
        /** @type {?} */
        const terminal = customData.tab === 'web' ? 'PC' : 'Mobile';
        /** @type {?} */
        let query = '?param=';
        /** @type {?} */
        let filter = `{"terminal":"${terminal}"`;
        if (searchValue.searchValue) {
            if (searchValue.searchField === '*') {
                filter += `,"allColumns":"${searchValue.searchValue}"`;
            }
            else {
                filter += `,"${searchValue.searchField}":"${searchValue.searchValue}"`;
            }
        }
        filter += '}';
        filter = encodeURIComponent(filter);
        query += (filter + `&pageNum=${index}&pageSize=${size}`);
        url += query;
        return this.http.get(url).pipe(map((/**
         * @param {?} re
         * @return {?}
         */
        (re) => ({
            items: re.content,
            total: re.totalElements,
            pageInfo: {
                pageIndex: re.pageable.pageNumber + 1,
                pageSize: re.pageable.pageSize,
                pageList: [20, 40, 60, 80, 100],
                enablePager: true
            }
        }))));
    }
}
FormFormatService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FormFormatService.ctorParameters = () => [
    { type: HttpService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    FormFormatService.prototype.http;
}
