// @gsp-cmp/ubcc-param-vo/esm2015/lib/vo/serializer/serializer-context/ubcc-serializer-context.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/serializer-context/ubcc-serializer-context.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DefaultOperationSerializerFactory, DefaultPropertySerializerFactory } from '@ecp-caf/common-structure';
import { UbccParamSerializerFactory } from '../factory/ubcc-param-serializer-factory';
import { UbccDataTypeSerializerFactory } from '../factory/ubcc-data-type-serializer-factory';
import { UbccParamConstraintSerializerFactory } from '../factory/ubcc-param-constraint-serializer-factory';
export class UbccParamSerializerContext {
    /**
     * @return {?}
     */
    getDataTypeSerializerFactory() {
        if (!this.dataTypeFactory) {
            this.dataTypeFactory = new UbccDataTypeSerializerFactory();
        }
        return this.dataTypeFactory;
    }
    /**
     * @return {?}
     */
    getParameterSerializerFactory() {
        if (!this.parameterFactory) {
            this.parameterFactory = new UbccParamSerializerFactory();
        }
        return this.parameterFactory;
    }
    /**
     * @return {?}
     */
    getOperationSerializerFactory() {
        if (!this.operationFactory) {
            this.operationFactory = new DefaultOperationSerializerFactory();
        }
        return this.operationFactory;
    }
    /**
     * @return {?}
     */
    getPropertySerializerFactory() {
        if (!this.propertyFactory) {
            this.propertyFactory = new DefaultPropertySerializerFactory();
        }
        return this.propertyFactory;
    }
    /**
     * @return {?}
     */
    getConstraintSerializerFactory() {
        if (!this.constraintFactory) {
            this.constraintFactory = new UbccParamConstraintSerializerFactory();
        }
        return this.constraintFactory;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    UbccParamSerializerContext.prototype.dataTypeFactory;
    /**
     * @type {?}
     * @private
     */
    UbccParamSerializerContext.prototype.parameterFactory;
    /**
     * @type {?}
     * @private
     */
    UbccParamSerializerContext.prototype.operationFactory;
    /**
     * @type {?}
     * @private
     */
    UbccParamSerializerContext.prototype.propertyFactory;
    /**
     * @type {?}
     * @private
     */
    UbccParamSerializerContext.prototype.constraintFactory;
}
