// @gsp-svc/cloudprint/esm5/lib/entitys/datasourceoptions/retrieveParam.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 这个类以后不用了，但是不能删，财务已经使用
 */
var RetrieveParam = /** @class */ (function () {
    function RetrieveParam() {
    }
    return RetrieveParam;
}());
export { RetrieveParam };
if (false) {
    /** @type {?} */
    RetrieveParam.prototype.NeedLock;
    /** @type {?} */
    RetrieveParam.prototype.NodeSortInfos;
    /** @type {?} */
    RetrieveParam.prototype.CacheType;
    /** @type {?} */
    RetrieveParam.prototype.ForceFromRepository;
}
var NodeSortInfos = /** @class */ (function () {
    function NodeSortInfos() {
    }
    return NodeSortInfos;
}());
export { NodeSortInfos };
if (false) {
    /** @type {?} */
    NodeSortInfos.prototype.NodeCode;
    /** @type {?} */
    NodeSortInfos.prototype.SortConditions;
}
var SortConditions = /** @class */ (function () {
    function SortConditions() {
    }
    return SortConditions;
}());
export { SortConditions };
if (false) {
    /** @type {?} */
    SortConditions.prototype.SortType;
    /** @type {?} */
    SortConditions.prototype.SortField;
}
/** @enum {number} */
var RequestedBufferType = {
    //
    // 摘要:
    //     本次请求中的数据。
    //
    // 言论：
    //     如果本次请求执行成功,提交到Inspur.Gsp.Bef.Api.Parameter.RequestedBufferType.TransactionalBuffer,否则被丢弃
    CurrentData: 0,
    //
    // 摘要:
    //     包含会话过程中已经接受的变更数据
    TransactionalBuffer: 1,
    //
    // 摘要:
    //     修改之前的老值,用于比较变更,获取差异或触发事件。会话过程中保持与数据库中一致
    Original: 2,
};
export { RequestedBufferType };
RequestedBufferType[RequestedBufferType.CurrentData] = 'CurrentData';
RequestedBufferType[RequestedBufferType.TransactionalBuffer] = 'TransactionalBuffer';
RequestedBufferType[RequestedBufferType.Original] = 'Original';
/** @enum {number} */
var SortType = {
    Asc: 0,
    Desc: 1,
};
export { SortType };
SortType[SortType.Asc] = 'Asc';
SortType[SortType.Desc] = 'Desc';
