// @gsp-sys/sysmgr-ui/esm5/lib/sys-manager-ui/group-nav/group.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { SysCommonUtilService, SysDataUtil } from '@gsp-sys/sysmgr-common';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "@gsp-sys/sysmgr-common";
var GroupService = /** @class */ (function () {
    function GroupService(http, sysUtil) {
        this.http = http;
        this.sysUtil = sysUtil;
    }
    /**
     * @private
     * @param {?} type
     * @return {?}
     */
    GroupService.prototype.getUrlByGroupType = /**
     * @private
     * @param {?} type
     * @return {?}
     */
    function (type) {
        /** @type {?} */
        var url = '';
        switch (type) {
            case 'user':
                url = this.sysUtil.userGroupWebApi;
                break;
            case 'position':
                url = this.sysUtil.posGroupWebApi;
                break;
            case 'role':
                url = this.sysUtil.roleGroupWebApi;
                break;
            case 'org':
                url = this.sysUtil.sysOrgWebApi;
                break;
            // 默认用户分组
            default:
                url = this.sysUtil.userGroupWebApi;
                break;
        }
        return url;
    };
    /**
     * 根据级数获取组织
     * @param layer 级数
     */
    /**
     * 根据级数获取组织
     * @param {?} layer 级数
     * @param {?} type
     * @param {?} filter
     * @return {?}
     */
    GroupService.prototype.getByLayer = /**
     * 根据级数获取组织
     * @param {?} layer 级数
     * @param {?} type
     * @param {?} filter
     * @return {?}
     */
    function (layer, type, filter) {
        /** @type {?} */
        var url = this.getUrlByGroupType(type);
        if (!filter) {
            filter = {};
        }
        filter.layer = layer;
        filter.parentId = null;
        SysDataUtil.deleteEmpty(filter);
        /** @type {?} */
        var opt = {
            headers: this.sysUtil.getHeader().headers,
            params: { param: JSON.stringify(filter) }
        };
        return this.http.get(url, opt);
    };
    /**
     * 根据级数获取组织
     * @param parentId 父节点
     */
    /**
     * 根据级数获取组织
     * @param {?} parentId 父节点
     * @param {?} type
     * @param {?} filter
     * @return {?}
     */
    GroupService.prototype.getByParentId = /**
     * 根据级数获取组织
     * @param {?} parentId 父节点
     * @param {?} type
     * @param {?} filter
     * @return {?}
     */
    function (parentId, type, filter) {
        /** @type {?} */
        var url = this.getUrlByGroupType(type);
        if (!filter) {
            filter = {};
        }
        filter.layer = null;
        filter.parentId = parentId;
        SysDataUtil.deleteEmpty(filter);
        /** @type {?} */
        var opt = {
            headers: this.sysUtil.getHeader().headers,
            params: { param: JSON.stringify(filter) }
        };
        return this.http.get(url, opt);
    };
    /**
     * 根据级数获取组织
     * @param parentId 父节点
     */
    /**
     * 根据级数获取组织
     * @param {?} type
     * @param {?} filter
     * @return {?}
     */
    GroupService.prototype.searchData = /**
     * 根据级数获取组织
     * @param {?} type
     * @param {?} filter
     * @return {?}
     */
    function (type, filter) {
        /** @type {?} */
        var url = this.getUrlByGroupType(type);
        if (!filter) {
            filter = {};
        }
        filter.layer = null;
        filter.parentId = null;
        SysDataUtil.deleteEmpty(filter);
        /** @type {?} */
        var opt = {
            headers: this.sysUtil.getHeader().headers,
            params: { param: JSON.stringify(filter) }
        };
        return this.http.get(url, opt);
    };
    GroupService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    GroupService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: SysCommonUtilService }
    ]; };
    /** @nocollapse */ GroupService.ngInjectableDef = i0.defineInjectable({ factory: function GroupService_Factory() { return new GroupService(i0.inject(i1.HttpClient), i0.inject(i2.SysCommonUtilService)); }, token: GroupService, providedIn: "root" });
    return GroupService;
}());
export { GroupService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    GroupService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    GroupService.prototype.sysUtil;
}
