// @farris/discussion-group/esm5/lib/discussion-editor/discussion-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, ElementRef, Renderer2, Output, EventEmitter, ViewEncapsulation, ChangeDetectorRef, Injector } from '@angular/core';
import { NotifyService } from '@farris/ui-notify';
import { LocaleService } from '@farris/ui-locale';
import { MsgInfo } from './discussion-editor.interface';
import { HttpClient } from '@angular/common/http';
import { of } from 'rxjs';
var DiscussionEditorComponent = /** @class */ (function () {
    function DiscussionEditorComponent(renderer, elementRef, cd, notifyService, injector, http) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.cd = cd;
        this.notifyService = notifyService;
        this.injector = injector;
        this.http = http;
        this.personnelsPrimaryKey = 'id';
        this.personnelsDisplayKey = 'name';
        this.replyPersonnelsDisplayKey = 'userName';
        // @Input() personnels: Array<any> = [];
        // @Input() permissionList: Array<any> = [
        //     { value: 'ALL', text: this.localeService.getValue('discussionGroup.all') },
        //     { value: 'RELATED', text: this.localeService.getValue('discussionGroup.related') }
        // ];
        this.cancelVisible = true;
        this.editHeight = 130;
        this.type = 'user';
        // 人事弹窗列表数据
        this._personnels = [];
        // 新建还是维护自定义角色
        // @Output() customGroupChange = new EventEmitter<any>();
        // 对外暴露数据
        this.selectionsChange = new EventEmitter();
        // 选中的行
        this.lineDataChange = new EventEmitter();
        // 对外
        this.valueChange = new EventEmitter();
        /**
         * 附件预览
         */
        this.filePreview = new EventEmitter();
        /**
         * 附件删除
         */
        this.fileRemove = new EventEmitter();
        this.fileUploadDone = new EventEmitter();
        this.personnelSearch = new EventEmitter();
        this.getOugUsers = new EventEmitter();
        // @ViewChild('searchInput') searchInput;
        // 审批意见
        this.textValue = '';
        // 暂存人员信息
        this.tempPersonnelsValue = '';
        // 暂存部门
        this.tempSectionValue = '';
        // 搜索人事管理弹窗
        this.personModalVisible = false;
        // 选择要发送的人员列表
        this.selectedPersonnels = [];
        // 选择要发送的部门
        this.selectedSection = [];
        // 暂存人员信息，用于搜索
        this.copyPersonnels = [];
        this.innerPersonnels = [];
        this.relativeVisible = false;
        //上传附件是否显示
        this.attachFilesModalVisible = false;
        this.pageSize = 20;
        this.searchPersonnelList = {};
        this.showSearchList = false;
        this.permissionList = [];
        this.groupIcon = '<span class="f-icon f-icon-search"></span>';
        this.localeService = this.injector.get(LocaleService);
    }
    Object.defineProperty(DiscussionEditorComponent.prototype, "personnels", {
        get: /**
         * @return {?}
         */
        function () {
            return this._personnels;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val) {
                this._personnels = val;
                this.innerPersonnels = this._personnels;
                this.copyPersonnels = this._personnels;
                // this.copyPersonnels = this._personnels;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DiscussionEditorComponent.prototype, "replyUser", {
        get: /**
         * @return {?}
         */
        function () {
            return this._replyUser;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val) {
                this._replyUser = val;
                if (this._replyUser['id']) {
                    this.editorFocus();
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DiscussionEditorComponent.prototype, "attachFiles", {
        get: /**
         * @return {?}
         */
        function () {
            return this._attachFiles;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val) {
                this._attachFiles = val;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DiscussionEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var self = this;
        this.renderer.listen(document, 'click', (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            self.personModalVisible = false;
            self.relativeVisible = false;
        }));
        // this.copyPersonnels = this.personnels;
        this.el = this.elementRef.nativeElement;
        this.permissionList = [
            { value: 'ALL', text: this.localeService.getValue('discussionGroup.all') },
            { value: 'RELATED', text: this.localeService.getValue('discussionGroup.related') }
        ];
        this.permission = this.permissionList[0];
        this.options = { maxUploads: 3, maxFileSize: 10240, allowedContentTypes: ['.jpg', '.pdf'] };
        this.placeholder = this.localeService.getValue('discussionGroup.placeholder');
    };
    /**
     * 单个附件上传完成
     * @param ev
     */
    /**
     * 单个附件上传完成
     * @param {?} ev
     * @return {?}
     */
    DiscussionEditorComponent.prototype.fUploadDoneHandler = /**
     * 单个附件上传完成
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        this.fileUploadDone.emit(ev);
    };
    /**
     * 上传附件中单个附件删除完成
     * @param ev
     */
    /**
     * 上传附件中单个附件删除完成
     * @param {?} ev
     * @return {?}
     */
    DiscussionEditorComponent.prototype.fUploadRemovedHandler = /**
     * 上传附件中单个附件删除完成
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        // console.log('fUploadRemovedHandler');
        // console.log(ev);
    };
    /**附件列表中预览 */
    /**
     * 附件列表中预览
     * @param {?} info
     * @return {?}
     */
    DiscussionEditorComponent.prototype.filePreviewEventHandler = /**
     * 附件列表中预览
     * @param {?} info
     * @return {?}
     */
    function (info) {
        this.filePreview.emit(info);
    };
    /**附件列表中删除 */
    /**
     * 附件列表中删除
     * @param {?} file
     * @return {?}
     */
    DiscussionEditorComponent.prototype.fileRemoveEventHandler = /**
     * 附件列表中删除
     * @param {?} file
     * @return {?}
     */
    function (file) {
        this.fileRemove.emit(file);
    };
    // 文本框失去焦点触发
    // 文本框失去焦点触发
    /**
     * @param {?=} e
     * @return {?}
     */
    DiscussionEditorComponent.prototype.setTextValue = 
    // 文本框失去焦点触发
    /**
     * @param {?=} e
     * @return {?}
     */
    function (e) {
        if (e === void 0) { e = null; }
        if (e) {
            // console.log(e.target.innerHTML);
            this.tempTextValue = e.target.innerHTML;
            this.textValue = this.tempTextValue;
        }
        // if (this.tempTextValue) {
        //     this.textValue = this.tempTextValue
        // }
        if (this.tempPersonnelsValue) {
            this.textValue += this.tempPersonnelsValue;
        }
        if (this.tempSectionValue) {
            this.textValue += this.tempSectionValue;
        }
        this.tempTextValue = '';
        this.tempPersonnelsValue = '';
        this.tempSectionValue = '';
    };
    // 监听键盘事件, 主要是用于删除@人
    // 监听键盘事件, 主要是用于删除@人
    /**
     * @param {?} e
     * @return {?}
     */
    DiscussionEditorComponent.prototype.listenEditorValueChange = 
    // 监听键盘事件, 主要是用于删除@人
    /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        var _this = this;
        this.tempTextValue = e.target.innerHTML;
        /** @type {?} */
        var children = e.target.children;
        /** @type {?} */
        var childrenId = [];
        for (var i = 0; i < children.length; i++) {
            childrenId.push(children[i].id);
        }
        this.selectedPersonnels.forEach((/**
         * @param {?} personnel
         * @param {?} index
         * @return {?}
         */
        function (personnel, index) {
            if (!childrenId.includes(personnel[_this.personnelsPrimaryKey])) {
                _this.selectedPersonnels.splice(index, 1);
            }
        }));
        this.selectedSection.forEach((/**
         * @param {?} section
         * @param {?} index
         * @return {?}
         */
        function (section, index) {
            if (!childrenId.includes(section[_this.personnelsPrimaryKey])) {
                _this.selectedSection.splice(index, 1);
            }
        }));
        if (!this.tempTextValue) {
            this.tempTextValue = '';
        }
    };
    /**
     * 搜索人员
     */
    /**
     * 搜索人员
     * @return {?}
     */
    DiscussionEditorComponent.prototype.serachPersonnel = /**
     * 搜索人员
     * @return {?}
     */
    function () {
        var _this = this;
        // if (this.personnelText) {
        //     this.innerPersonnels = this.copyPersonnels.filter(item => {
        //         if(item[this.personnelsDisplayKey]){
        //             return item[this.personnelsDisplayKey].indexOf(this.personnelText) !== -1;
        //         }
        //     })
        // }
        // else {
        //     this.innerPersonnels = this.copyPersonnels;
        // }
        if (this.personnelText) {
            this.showSearchList = true;
            this.getSearchData(this.personnelText, 0).subscribe((/**
             * @param {?} d
             * @return {?}
             */
            function (d) {
                if (d.hasOwnProperty("users")) {
                    _this.searchPersonnelList = d;
                    _this.setPersonModalPosition();
                }
            }));
        }
        else {
            this.showSearchList = false;
        }
        // this.searchInput.nativeElement.blur();
        //暂时注释掉服务器端查询
        // this.personnelSearch.emit(this.personnelText);
    };
    /**
     * 搜索下一页
     */
    /**
     * 搜索下一页
     * @return {?}
     */
    DiscussionEditorComponent.prototype.getMoreSearchData = /**
     * 搜索下一页
     * @return {?}
     */
    function () {
        var _this = this;
        this.getSearchData(this.personnelText, this.searchPersonnelList.pageIndex + 1).subscribe((/**
         * @param {?} d
         * @return {?}
         */
        function (d) {
            if (d.hasOwnProperty("users")) {
                _this.searchPersonnelList.pageIndex = d['pageIndex'];
                _this.searchPersonnelList['users'] = tslib_1.__spread(_this.searchPersonnelList['users'], d['users']);
            }
        }));
    };
    /**
     * @param {?} text
     * @param {?} pageIndex
     * @param {?=} pageSize
     * @return {?}
     */
    DiscussionEditorComponent.prototype.getSearchData = /**
     * @param {?} text
     * @param {?} pageIndex
     * @param {?=} pageSize
     * @return {?}
     */
    function (text, pageIndex, pageSize) {
        if (pageSize === void 0) { pageSize = this.pageSize; }
        if (this.personSearchUrl) {
            /** @type {?} */
            var url = this.personSearchUrl + "?param=" + text + "&pageSize=" + pageSize + "&pageIndex=" + pageIndex;
            //console.log(url);
            return this.http.get(url);
            //return this.http.get('assets/data/discussionPerson3.json');
        }
        else {
            return of(true);
        }
    };
    /**
     * @param {?} item
     * @return {?}
     */
    DiscussionEditorComponent.prototype.personnelChecked = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        item.active = !item.active;
        this.innerPersonnels = this.innerPersonnels.map((/**
         * @param {?} o
         * @return {?}
         */
        function (o) { return (tslib_1.__assign({}, o)); }));
        this.cd.markForCheck();
        this.appendPersonnels();
    };
    // 增加@人员
    // 增加@人员
    /**
     * @return {?}
     */
    DiscussionEditorComponent.prototype.appendPersonnels = 
    // 增加@人员
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var selectedList = [];
        if (!this.showSearchList) {
            selectedList = this.innerPersonnels.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.active === true; }));
        }
        else {
            selectedList = this.searchPersonnelList['users'].filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.active === true; }));
        }
        if (selectedList.length) {
            this.appendPersonnel(selectedList);
        }
        this.resetPersonnels();
        this.setTextValue();
        this.personModalVisible = false;
    };
    //高级搜索人员添加
    //高级搜索人员添加
    /**
     * @param {?} listData
     * @return {?}
     */
    DiscussionEditorComponent.prototype.appendPersonnelsList = 
    //高级搜索人员添加
    /**
     * @param {?} listData
     * @return {?}
     */
    function (listData) {
        if (listData.length) {
            this.appendPersonnel(listData, true);
        }
        this.setTextValue();
    };
    //循环增加人员
    //循环增加人员
    /**
     * @param {?} listData
     * @param {?=} external
     * @return {?}
     */
    DiscussionEditorComponent.prototype.appendPersonnel = 
    //循环增加人员
    /**
     * @param {?} listData
     * @param {?=} external
     * @return {?}
     */
    function (listData, external) {
        var _this = this;
        if (external === void 0) { external = false; }
        listData.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (_this.selectedPersonnels.length && _this._isInArray(item[_this.personnelsPrimaryKey], _this.personnelsPrimaryKey, _this.selectedPersonnels)) {
                return;
            }
            else {
                //if (!external) {
                _this.tempPersonnelsValue += '<a name="personnel" contenteditable="false" class="at-user" id="' + item[_this.personnelsPrimaryKey] + '" userid="' + item[_this.personnelsPrimaryKey] + '">@' + item[_this.personnelsDisplayKey] + '</a>&nbsp;';
                // }
                // else {
                //     this.tempPersonnelsValue += '<a name="personnel" contenteditable="false" class="at-user" id="' + item[this.personnelsPrimaryKey] + '" userid="' + item[this.personnelsPrimaryKey] + '">@' + item[this.replyPersonnelsDisplayKey] + '</a>';
                // }
                _this.selectedPersonnels.push(item);
            }
        }));
    };
    /**
     * 添加部门
     * @param listData
     */
    /**
     * 添加部门
     * @param {?} listData
     * @return {?}
     */
    DiscussionEditorComponent.prototype.appendSectionList = /**
     * 添加部门
     * @param {?} listData
     * @return {?}
     */
    function (listData) {
        if (listData.length) {
            this.appendSection(listData);
        }
        this.setTextValue();
    };
    /**
     * @param {?} listData
     * @return {?}
     */
    DiscussionEditorComponent.prototype.appendSection = /**
     * @param {?} listData
     * @return {?}
     */
    function (listData) {
        var _this = this;
        listData.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (_this.selectedSection.length && _this._isInArray(item[_this.personnelsPrimaryKey], _this.personnelsPrimaryKey, _this.selectedSection)) {
                return;
            }
            else {
                _this.tempSectionValue += '<a name="org" contenteditable="false" class="at-org" id="' + item[_this.personnelsPrimaryKey] + '" orgid="' + item[_this.personnelsPrimaryKey] + '">@' + item['name'] + '</a>&nbsp;';
                _this.selectedSection.push(item);
            }
        }));
    };
    // 判断是否在数组
    // 判断是否在数组
    /**
     * @private
     * @param {?} value
     * @param {?} fieldInArray
     * @param {?} array
     * @return {?}
     */
    DiscussionEditorComponent.prototype._isInArray = 
    // 判断是否在数组
    /**
     * @private
     * @param {?} value
     * @param {?} fieldInArray
     * @param {?} array
     * @return {?}
     */
    function (value, fieldInArray, array) {
        if (!value || !fieldInArray) {
            return false;
        }
        return array.findIndex((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return value === item[fieldInArray]; })) === -1 ? false : true;
    };
    /**
     * 关闭人事管理弹窗
     */
    /**
     * 关闭人事管理弹窗
     * @return {?}
     */
    DiscussionEditorComponent.prototype.resetPersonnels = /**
     * 关闭人事管理弹窗
     * @return {?}
     */
    function () {
        this.showSearchList = false;
        this.personModalVisible = false;
        this.innerPersonnels = this.copyPersonnels;
        if (this.innerPersonnels.length) {
            this.innerPersonnels.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.active = false; }));
        }
        this.personnelText = '';
        // this.stopBubble(e);
    };
    /**
     * 打开人员管理
     * @param e 事件
     */
    /**
     * 打开人员管理
     * @param {?} e 事件
     * @return {?}
     */
    DiscussionEditorComponent.prototype.openModalPerson = /**
     * 打开人员管理
     * @param {?} e 事件
     * @return {?}
     */
    function (e) {
        var _this = this;
        this.personModalVisible = !this.personModalVisible;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.personModalVisible) {
                _this.setPersonModalPosition();
            }
        }), 0);
        // this.getPersonnelList();
        this.stopBubble(e);
    };
    /**
     * @return {?}
     */
    DiscussionEditorComponent.prototype.setPersonModalPosition = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var winH = window.innerHeight;
        /** @type {?} */
        var ModalBottom = this.personModal.nativeElement.getBoundingClientRect().bottom;
        if (winH < ModalBottom) {
            this.personModal.nativeElement.scrollIntoView(false);
        }
    };
    /**
     * @param {?} e
     * @return {?}
     */
    DiscussionEditorComponent.prototype.openModalPermission = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.relativeVisible = !this.relativeVisible;
        this.stopBubble(e);
    };
    /**
     * 打开上传附件
     */
    /**
     * 打开上传附件
     * @param {?} e
     * @return {?}
     */
    DiscussionEditorComponent.prototype.openModalFile = /**
     * 打开上传附件
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.attachFilesModalVisible = !this.attachFilesModalVisible;
        this.stopBubble(e);
    };
    // 阻止冒泡
    // 阻止冒泡
    /**
     * @param {?} e
     * @return {?}
     */
    DiscussionEditorComponent.prototype.stopBubble = 
    // 阻止冒泡
    /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (e && e.stopPropagation) {
            e.stopPropagation();
        }
        else {
            window.event.cancelBubble = true;
        }
    };
    /**
     * 提交评语
     */
    /**
     * 提交评语
     * @return {?}
     */
    DiscussionEditorComponent.prototype.submitApproval = /**
     * 提交评语
     * @return {?}
     */
    function () {
        if (!this.textValue) {
            /** @type {?} */
            var notEmptyText = this.localeService.getValue('discussionGroup.notEmpty');
            if (notEmptyText) {
                this.notifyService.error(notEmptyText);
            }
            return;
        }
        /** @type {?} */
        var editAttachFiles = [];
        if (this.attachFiles && this.attachFiles.length) {
            this.attachFiles.forEach((/**
             * @param {?} file
             * @return {?}
             */
            function (file) {
                /** @type {?} */
                var id = file['id'];
                /** @type {?} */
                var name = file['name'];
                /** @type {?} */
                var size = file['size'];
                /** @type {?} */
                var metadataId = file['extend']['metadataId'];
                /** @type {?} */
                var attachFile = {
                    id: id,
                    name: name,
                    size: size,
                    metadataId: metadataId
                };
                editAttachFiles.push(attachFile);
            }));
        }
        this.valueChange.emit({
            msgInfo: MsgInfo.Confirm,
            text: this.textValue,
            mailTos: this.selectedPersonnels,
            mailToSections: this.selectedSection,
            visibility: this.permission.value,
            parentId: (this.replyUser && this.replyUser.hasOwnProperty('id')) ? this.replyUser.id : null,
            attachFiles: editAttachFiles.length ? editAttachFiles : null
        });
        this.textValue = '';
        this.selectedPersonnels = [];
        this.selectedSection = [];
        this.attachFiles = [];
        this.replyUser = {};
    };
    /**
     * @return {?}
     */
    DiscussionEditorComponent.prototype.cancel = /**
     * @return {?}
     */
    function () {
        this.valueChange.emit({
            msgInfo: MsgInfo.Cancel,
            text: null,
            mailTos: [],
            mailToSections: [],
            visibility: null,
            parentId: null,
            attachFiles: null
        });
        this.textValue = '';
        this.selectedPersonnels = [];
        this.selectedSection = [];
        this.attachFiles = [];
        this.replyUser = {};
    };
    /**修改查看权限 */
    /**
     * 修改查看权限
     * @param {?} value
     * @return {?}
     */
    DiscussionEditorComponent.prototype.changePermission = /**
     * 修改查看权限
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.permission = this.permissionList.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.value === value; }));
        this.relativeVisible = false;
    };
    //获得焦点
    //获得焦点
    /**
     * @return {?}
     */
    DiscussionEditorComponent.prototype.editorFocus = 
    //获得焦点
    /**
     * @return {?}
     */
    function () {
        this.editor.nativeElement.focus();
    };
    // customGroupChangePar(event){
    //     this.customGroupChange.emit(event);
    // }
    /**
     * 高级人员点确认
     */
    // customGroupChangePar(event){
    //     this.customGroupChange.emit(event);
    // }
    /**
     * 高级人员点确认
     * @param {?} event
     * @return {?}
     */
    DiscussionEditorComponent.prototype.selectionsChangePar = 
    // customGroupChangePar(event){
    //     this.customGroupChange.emit(event);
    // }
    /**
     * 高级人员点确认
     * @param {?} event
     * @return {?}
     */
    function (event) {
        //this.resetPersonnels();
        if (event.data.users.length) {
            /** @type {?} */
            var userList_1 = [];
            event.data.users.forEach((/**
             * @param {?} user
             * @return {?}
             */
            function (user) {
                userList_1.push(user.data);
            }));
            this.appendPersonnelsList(userList_1);
        }
        if (event.data.section.length) {
            /** @type {?} */
            var sectionList_1 = [];
            event.data.section.forEach((/**
             * @param {?} sec
             * @return {?}
             */
            function (sec) {
                sectionList_1.push(sec.data);
            }));
            this.appendSectionList(sectionList_1);
        }
        this.selectionsChange.emit(event);
    };
    /**高级人员中选中某行 */
    /**
     * 高级人员中选中某行
     * @param {?} event
     * @return {?}
     */
    DiscussionEditorComponent.prototype.lineDataChangePar = /**
     * 高级人员中选中某行
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.lineDataChange.emit(event);
    };
    /**获得占位头像 */
    /**
     * 获得占位头像
     * @param {?} item
     * @return {?}
     */
    DiscussionEditorComponent.prototype.getAvatar = /**
     * 获得占位头像
     * @param {?} item
     * @return {?}
     */
    function (item) {
        if (item && item[this.personnelsDisplayKey]) {
            /** @type {?} */
            var str = item[this.personnelsDisplayKey];
            return str.substring(str.length - 2, str.length);
        }
        else {
            return '';
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DiscussionEditorComponent.prototype.ougUsers = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.getOugUsers.emit(event);
    };
    DiscussionEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-discussion-editor',
                    template: "<div class=\"farris-discussion-group-reply-tip\" *ngIf=\"replyUser && replyUser.id\">\r\n    {{ 'discussionGroup.reply' | locale }} &nbsp;\r\n    <span class=\"farris-discussion-group-reply-tip-name\">\r\n        {{replyUser[replyPersonnelsDisplayKey]}}\r\n    </span>\r\n    \uFF1A\r\n</div>\r\n<div class=\"f-discussion-group-edit\">\r\n    <div class=\"f-discussion-group-edit-container\" [ngStyle]=\"{'height': editHeight + 'px'}\">\r\n        <div class=\"textarea-editor\" #editor (keyup)=\"listenEditorValueChange($event)\" (blur)=\"setTextValue($event)\"\r\n            contenteditable=\"true\" [innerHTML]=\"textValue | safeHtml\"></div>\r\n    </div>\r\n    <!-- \u9644\u4EF6\u90E8\u5206 <div class=\"f-discussion-group-edit-files\" *ngIf=\"attachFiles && attachFiles.length\">\r\n        <f-file-preview class=\"d-flex flex-wrap\" [itemCls]=\"'demo-view-item'\" [fileInfos]=\"attachFiles\"\r\n            (filePreviewEvent)=\"filePreviewEventHandler($event)\" (fileRemoveEvent)=\"fileRemoveEventHandler($event)\"\r\n            #filePreview>\r\n            <ng-template filePreviewListTypeTemplate let-data let-readonly=\"readonly\">\r\n                <div class=\"ffilepreview--item-detail\">\r\n                    <div class=\"ffilepreview--item-icon\">\r\n                        <span class=\"ffilepreview--filetype-icon\" [ngClass]=\"filePreview.getFileTypeClassName(data.name)\"></span>\r\n                    </div>\r\n                    <div class=\"ffilepreview--item-content\">\r\n                        <a class=\"item-content--title\" [title]=\"data.name\"\r\n                            (click)=\"filePreview.filePreviewHandler($event,data)\">{{data.name}}</a>\r\n                        <div class=\"item-content--toolbar f-text-light\">\r\n                            <button class=\"btn btn-link\" (click)=\"filePreview.filePreviewHandler($event,data)\">\r\n                                <span class=\"f-icon f-icon-preview\"></span>\u9884\u89C8\r\n                            </button>\r\n                            <button class=\"btn btn-link\" (click)=\"filePreview.fileRemoveHandler($event,data)\">\r\n                                <span class=\"f-icon f-icon-delete\"></span>\u5220\u9664\r\n                            </button>\r\n                        </div>\r\n                    </div>\r\n                </div>\r\n            </ng-template>\r\n        </f-file-preview>\r\n    </div> -->\r\n    <div class=\"f-discussion-group-edit-footer\">\r\n        <div class=\"f-discussion-group-edit-toolbar\">\r\n            <!-- <div class=\"f-discussion-group-edit-toolbar-item f-discussion-group-edit-toolbar-file\" (click)=\"openModalFile($event)\">\r\n                <div class=\"toolbar-item-text\">\r\n                    <span class=\"toolbar-icon f-icon f-icon-attachment\"></span>\r\n                    <span class=\"toolbar-text\">\u9644\u4EF6</span>\r\n                </div>\r\n            </div> -->\r\n            <div class=\"f-discussion-group-edit-toolbar-item f-discussion-group-edit-toolbar-about\">\r\n                <div class=\"toolbar-item-text\" (click)=\"openModalPerson($event)\">\r\n                    <span class=\"toolbar-icon\">@</span>\r\n                    <span class=\"toolbar-text\">{{ 'discussionGroup.colleague' | locale }}</span>\r\n                </div>\r\n                <div class=\"f-discussion-group-about-dropdown\" *ngIf=\"personModalVisible\" (click)=\"stopBubble($event)\" #personModal>\r\n                    <div class=\"f-discussion-group-about-search\">\r\n                        <!-- bugNum:480647 -->\r\n                        <input-group [placeholder]=\"placeholder\" [groupText]=\"groupIcon\" [(ngModel)]=\"personnelText\" (enterHandle)=\"serachPersonnel()\" (clickHandle)=\"serachPersonnel()\" (valueChange)=\"serachPersonnel()\" (clear)=\"serachPersonnel()\"></input-group>\r\n                    </div>\r\n                    <div class=\"f-discussion-group-about-content\">\r\n                        <div class=\"f-discussion-group-about-used\" [hidden]=\"showSearchList\">\r\n                            <ul class=\"f-discussion-group-about-dropdown-list\">\r\n                                <ng-template [ngIf]=\"innerPersonnels.length\">\r\n                                    <li *ngFor=\"let item of innerPersonnels\" (click)=\"item.active = !item.active\" class=\"d-flex f-discussion-group-about-dropdown-list-item\"\r\n                                        [class.active]=\"item.active\">\r\n                                        <span class=\"f-icon f-icon-check\"></span>\r\n                                        <img [src]=\"item.imgData\" class=\"about-list-item-avatar\" *ngIf=\"item.imgData\"/>\r\n                                        <div class=\"about-list-item-avatar-tip\" *ngIf=\"!item.imgData\" [innerHTML]=\"getAvatar(item)\"></div>\r\n                                        <div class=\"f-discussion-group-about-dropdown-list-detail f-utils-fill\">\r\n                                            <div class=\"about-list-detail-text\">{{item[personnelsDisplayKey]}}</div>\r\n                                            <div class=\"about-list-detail-subtext\">{{item.email}}</div>\r\n                                        </div>\r\n                                    </li>\r\n                                </ng-template>\r\n                                <ng-template [ngIf]=\"!innerPersonnels.length\">\r\n                                    <li class=\"f-discussion-group-about-dropdown-list-item about-dropdown-list-item-empty\">{{ 'discussionGroup.emptyMessage' | locale }}</li>\r\n                                </ng-template>\r\n                            </ul>        \r\n                        </div>\r\n                        <div class=\"f-discussion-group-about-searchlist\" [hidden]=\"!showSearchList\">\r\n                            <ul class=\"f-discussion-group-about-dropdown-list\">\r\n                                <ng-template [ngIf]=\"searchPersonnelList && searchPersonnelList.users && searchPersonnelList.users.length\">\r\n                                    <li *ngFor=\"let item of searchPersonnelList['users']\" (click)=\"item.active = !item.active\" class=\"d-flex f-discussion-group-about-dropdown-list-item\"\r\n                                        [class.active]=\"item.active\">\r\n                                        <span class=\"f-icon f-icon-check\"></span>\r\n                                        <img [src]=\"item.imgData\" class=\"about-list-item-avatar\" *ngIf=\"item.imgData\"/>\r\n                                        <div class=\"about-list-item-avatar-tip\" *ngIf=\"!item.imgData\" [innerHTML]=\"getAvatar(item)\"></div>\r\n                                        <div class=\"f-discussion-group-about-dropdown-list-detail f-utils-fill\">\r\n                                            <div class=\"about-list-detail-text\">{{item[personnelsDisplayKey]}}</div>\r\n                                            <!-- <div class=\"about-list-detail-subtext\">{{item.email}}</div> -->\r\n                                            <div class=\"about-list-detail-subtext\">\r\n                                                <span [innerHTML]=\"item | getOrgText:sectionData\"></span>\r\n                                            </div>\r\n                                        </div>\r\n                                    </li>\r\n                                    <div class=\"f-discussion-person-list-page\">\r\n                                        <span class=\"section-page-more section-page-text\" *ngIf=\"searchPersonnelList.pageIndex < searchPersonnelList.pageCount-1\" (click)=\"getMoreSearchData()\">{{ 'discussionGroup.viewMore' | locale }}</span>\r\n                                    </div>\r\n                                </ng-template>\r\n                                <ng-template [ngIf]=\"searchPersonnelList && searchPersonnelList.users && !searchPersonnelList.users.length\">\r\n                                    <li class=\"f-discussion-group-about-dropdown-list-item\">{{ 'discussionGroup.emptyMessage' | locale }}</li>\r\n                                </ng-template>\r\n                            </ul>\r\n                        </div>\r\n                    </div>\r\n                    <div class=\"f-about-list-btns d-flex\">\r\n                        <div class=\"f-about-list-btns-left\">\r\n                            <farris-discussion-personnel \r\n                            [sectionData]=\"sectionData\"\r\n                            [orgUrl]=\"orgUrl\"\r\n                            [personSearchUrl]=\"personSearchUrl\"\r\n                            (selectionsChange)=\"selectionsChangePar($event)\"\r\n                            (lineDataChange)=\"lineDataChangePar($event)\"\r\n                            (ougUsers)=\"ougUsers($event)\"\r\n                            ></farris-discussion-personnel>\r\n                        </div>\r\n                        <div class=\"f-about-list-btns-right\">\r\n                            <button class=\"btn btn-secondary\" (click)=\"resetPersonnels()\">{{ 'discussionGroup.cancel' | locale }}</button>\r\n                            <button class=\"btn btn-primary\" (click)=\"appendPersonnels()\">{{ 'discussionGroup.confirm' | locale }}</button>\r\n                        </div>\r\n                        \r\n                    </div>\r\n                </div>\r\n            </div>\r\n            <!-- \u6682\u65F6\u6CE8\u91CA20210407 -->\r\n            <!-- <div class=\"f-discussion-group-edit-toolbar-item f-discussion-group-edit-toolbar-dropdown\">\r\n                <div class=\"toolbar-dropdown-text\" (click)=\"openModalPermission($event)\">\r\n                    <span class=\"toolbar-text\">{{permission.text}}</span>\r\n                    <span class=\"f-icon f-icon-collapse toolbar-icon\"></span>\r\n                </div>\r\n                <div class=\"toolbar-dropdown-list dropdownlist\" *ngIf=\"relativeVisible\">\r\n                    <ul *ngIf=\"permissionList && permissionList.length\">\r\n                        <li *ngFor=\"let item of permissionList\" class=\"dropdown-item\" [class.active]=\"permission?.value === item.value\"\r\n                            (click)=\"changePermission(item.value)\">{{item.text}}</li>\r\n                    </ul>\r\n                </div>\r\n            </div> -->\r\n        </div>\r\n        <div class=\"f-discussion-group-edit-btns\">\r\n            <button class=\"btn btn-link f-discussion-cancel\" *ngIf=\"cancelVisible\" (click)=\"cancel()\">{{ 'discussionGroup.cancel' | locale }}</button>\r\n            <button class=\"btn btn-primary\" (click)=\"submitApproval()\">{{ 'discussionGroup.submit' | locale }}</button>\r\n        </div>\r\n    </div>\r\n</div>\r\n<!-- <div class=\"f-discussion-group-file-upload\" [hidden]=\"!attachFilesModalVisible\">\r\n    <f-file-upload\r\n        (fUploadDoneEvent)=\"fUploadDoneHandler($event)\"\r\n        (fUploadRemovedEvent)=\"fUploadRemovedHandler($event)\">\r\n    </f-file-upload>\r\n</div> -->",
                    encapsulation: ViewEncapsulation.None,
                    styles: [":host{display:block}.f-discussion-group-edit{border:1px solid #dcdcdc;background:#fff}.f-discussion-group-edit .f-discussion-group-edit-container{width:100%;position:relative}.f-discussion-group-edit .f-discussion-group-edit-container .textarea-editor{width:100%;height:100%;padding:8px;margin:0 auto;outline:0;overflow-y:auto;word-break:break-all;white-space:normal;resize:none}.f-discussion-group-edit .f-discussion-group-edit-container .textarea-editor a{color:#3487e3}.f-discussion-group-edit .f-discussion-group-edit-footer{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;-webkit-box-align:center;align-items:center;flex-wrap:wrap;-webkit-box-pack:justify;justify-content:space-between;padding:10px 14px;background:#f7f7f7}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;flex-wrap:wrap}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-item{cursor:pointer;color:rgba(0,0,0,.65)}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-item .toolbar-item-text{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;line-height:22px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-item .toolbar-text{margin-left:2px;font-size:14px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-item .toolbar-icon{color:#667580;font-size:16px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-file{margin-right:12px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about{position:relative;margin-right:32px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown{position:absolute;top:calc(100% + 6px);left:0;width:300px;padding-top:16px;z-index:999;background:#fff;box-shadow:0 2px 8px 0 rgba(0,0,0,.15);border-radius:2px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list{max-height:282px;margin-top:10px;overflow-y:auto}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list::-webkit-scrollbar{width:7px;height:7px;background-color:#8e8e8e}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list::-webkit-scrollbar-track{border-radius:0;background-color:#fff;border:0;background-clip:padding-box;border-right:0}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list::-webkit-scrollbar-thumb{border-radius:3px;background-color:#dbdbdb;border:0;background-clip:content-box;opacity:.6;-webkit-transform:rotate(90deg);transform:rotate(90deg)}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item{padding:3px 14px;list-style:none;-webkit-box-align:center;align-items:center}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item .f-discussion-group-about-dropdown-list-detail{margin-left:6px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item .f-discussion-group-about-dropdown-list-detail .about-list-detail-text{line-height:20px;font-size:12px;color:#0a131a;margin:0 auto;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item .f-discussion-group-about-dropdown-list-detail .about-list-detail-subtext{font-size:12px;color:#8f9ca6;line-height:20px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item .f-icon-check{border:1px solid #d8d9d9;color:#fff;font-size:15px;border-radius:2px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item .f-icon-check::before{content:''}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item.active .f-icon-check{border:none;background:#4c91ff}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item.active .f-icon-check::before{content:'\\e118'}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item .about-list-item-avatar{display:block;width:30px;height:30px;margin-left:6px;border-radius:50%;overflow:hidden}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item .about-list-item-avatar-tip{width:32px;height:32px;margin-left:6px;font-size:12px;color:#fff;text-align:center;line-height:32px;border-radius:50%;background-color:#4796ff}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item:hover{background:#e6f7ff}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item.about-dropdown-list-item-empty{cursor:default}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-dropdown-list-item.about-dropdown-list-item-empty:hover{background:#fff}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-discussion-group-about-search{padding:0 14px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-about-list-btns{-webkit-box-pack:justify;justify-content:space-between;background:#faf9f9;border-top:1px solid #efefef;padding:10px 15px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-about-list-btns .btn{margin-right:10px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-about .f-discussion-group-about-dropdown .f-about-list-btns .btn:last-child{margin-right:0}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-dropdown{position:relative}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-dropdown .toolbar-dropdown-text{cursor:pointer}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-dropdown .toolbar-icon{color:rgba(0,0,0,.25)}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-dropdown .toolbar-text{margin-left:0}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-dropdown .toolbar-dropdown-list{position:absolute;left:0;top:calc(100% + 6px);z-index:999;padding:2px 0;background:#fff;box-shadow:0 2px 8px 0 rgba(0,0,0,.15);border-radius:2px}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-dropdown .toolbar-dropdown-list ul{max-width:220px;overflow-y:auto}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-dropdown .toolbar-dropdown-list ul li{font:14px/28px Microsoft yahei;padding:0 15px 0 10px;word-break:keep-all}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-toolbar .f-discussion-group-edit-toolbar-dropdown .toolbar-dropdown-list ul li:hover{background:#e6f7ff}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-btns{flex-shrink:0}.f-discussion-group-edit .f-discussion-group-edit-footer .f-discussion-group-edit-btns .f-discussion-cancel{margin-right:12px;font-size:14px;color:rgba(0,0,0,.75)}.f-discussion-group-edit .f-discussion-group-edit-files{padding:10px 14px 0}.f-discussion-group-edit .f-discussion-group-edit-files .ffilepreview--filetype-icon{width:32px;height:32px}.farris-discussion-group-reply-tip{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;margin-bottom:10px;font-size:14px;color:rgba(0,0,0,.75)}.f-discussion-group-file-upload{margin-top:10px}.f-discussion-dialog-content{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;-webkit-box-pack:justify;justify-content:space-between;height:100%;padding:14px;overflow:hidden}.f-discussion-dialog-content-left{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;width:600px;height:100%;padding:0}.f-discussion-dialog-content-left .f-discussion-personnel-search{flex-shrink:0;position:relative;margin-bottom:7px}.f-discussion-dialog-content-left .f-discussion-tab-content{border:1px solid #dcdcdc;border-radius:2px;padding:12px 0 7px}.f-discussion-dialog-content-left .f-discussion-tab-content .f-discussion-personnel-content .f-discussion-personnel-crumbs{flex-shrink:0;position:relative;width:100%;height:32px;padding:0 12px;overflow:hidden}.f-discussion-dialog-content-left .f-discussion-tab-content .f-discussion-personnel-content .f-discussion-personnel-crumbs .crumbs-list{width:100%;overflow:hidden}.f-discussion-dialog-content-left .f-discussion-tab-content .f-discussion-personnel-content .f-discussion-personnel-crumbs .crumbs-list-item{display:-webkit-box;display:flex;flex-wrap:nowrap;flex-shrink:0;-webkit-box-align:center;align-items:center;line-height:20px;font-size:14px;color:rgba(0,0,0,.85);cursor:pointer}.f-discussion-dialog-content-left .f-discussion-tab-content .f-discussion-personnel-content .f-discussion-personnel-crumbs .crumbs-list-item .crumbs-text{padding:0 2px;text-overflow:ellipsis;max-width:80px;max-height:20px;white-space:nowrap;overflow:hidden}.f-discussion-dialog-content-left .f-discussion-tab-content .f-discussion-personnel-content .f-discussion-personnel-crumbs .crumbs-list-item:first-child .crumbs-text{padding-left:0}.f-discussion-dialog-content-left .f-discussion-tab-content .f-discussion-personnel-content .f-discussion-personnel-crumbs .crumbs-list-item.crumbs-list-item-disabled{color:rgba(0,0,0,.55)}.f-discussion-dialog-content-left .f-discussion-tab-content .f-discussion-personnel-content .f-discussion-personnel-crumbs .crumbs-list-ellipsis{-webkit-box-align:center;align-items:center;position:absolute;left:12px;top:0;width:30px;height:32px;line-height:32px}.f-discussion-dialog-content-left .f-discussion-tab-content .f-discussion-personnel-content .f-discussion-personnel-crumbs .crumbs-list-padding{padding-left:30px}.f-discussion-dialog-content-left .f-discussion-tab-content .f-discussion-personnel-content.f-discussion-personnel-content-hide{display:none!important}.f-discussion-dialog-content-right{height:100%;margin-left:14px;padding:0;border:1px solid #dcdcdc;border-radius:2px}.f-discussion-dialog-content-right .f-discussion-selected-personnel-header{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;-webkit-box-pack:justify;justify-content:space-between;height:36px;padding:0 14px;background:#fafafa}.f-discussion-dialog-content-right .f-discussion-selected-personnel-header-text{font-size:14px;line-height:20px}.f-discussion-dialog-content-right .f-discussion-selected-personnel-content{padding:8px 0}.f-discussion-dialog-content-right .f-discussion-selected-personnel-content .f-discussion-selected-list .f-discussion-selected-item{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;-webkit-box-pack:justify;justify-content:space-between;padding:6px 12px}.f-discussion-dialog-content-right .f-discussion-selected-personnel-content .f-discussion-selected-list .f-discussion-selected-item-text{line-height:20px}.f-discussion-dialog-content-right .f-discussion-selected-personnel-content .f-discussion-selected-list .f-discussion-selected-item-text .f-discussion-selected-name{font-size:14px;color:rgba(0,0,0,.85)}.f-discussion-dialog-content-right .f-discussion-selected-personnel-content .f-discussion-selected-list .f-discussion-selected-item-tip{font-size:12px;color:rgba(0,0,0,.35);line-height:20px}.f-discussion-dialog-content-right .f-discussion-selected-personnel-content .f-discussion-selected-list .f-discussion-selected-item-remove{visibility:hidden;color:#388fff;cursor:pointer}.f-discussion-dialog-content-right .f-discussion-selected-personnel-content .f-discussion-selected-list .f-discussion-selected-item:hover{background:#f3fbff}.f-discussion-dialog-content-right .f-discussion-selected-personnel-content .f-discussion-selected-list .f-discussion-selected-item:hover .f-discussion-selected-item-remove{visibility:visible}.f-discussion-dialog-content .f-discussion-selected-item-remove{visibility:hidden}.f-discussion-dialog-content .f-discussion-personnel-item{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;padding:3px 14px 3px 3px;cursor:pointer}.f-discussion-dialog-content .f-discussion-personnel-item-avatar{flex-shrink:0;width:36px;height:36px;margin-right:5px;border-radius:50%;overflow:hidden}.f-discussion-dialog-content .f-discussion-personnel-item-avatar img{display:block;width:100%}.f-discussion-dialog-content .f-discussion-personnel-item-text{flex-shrink:1;-webkit-box-flex:1;flex-grow:1;flex-basis:0;padding-right:10px;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}.f-discussion-dialog-content .f-discussion-personnel-item-text .f-discussion-personnel-name{line-height:20px;font-size:12px;color:#0a131a}.f-discussion-dialog-content .f-discussion-personnel-item-text .f-discussion-personnel-mail{line-height:20px;font-size:12px;color:#8f9ca6}.f-discussion-dialog-content .f-discussion-personnel-item-remove{display:none}.f-discussion-dialog-content .f-discussion-personnel-item:hover{background:#f3fbff}.f-discussion-dialog-content .f-discussion-personnel-item:hover .f-discussion-personnel-item-remove{display:block}.f-discussion-dialog-content .f-discussion-personnel-item.active{background:#def1ff}.f-discussion-dialog-content .f-discussion-personnel-item.selected{background:#fff;cursor:not-allowed}.f-discussion-dialog-content .f-discussion-personnel-item.selected .f-discussion-personnel-mail,.f-discussion-dialog-content .f-discussion-personnel-item.selected .f-discussion-personnel-name{color:#c1c1c1}.f-discussion-dialog-content .f-discussion-personnel-empty{padding:3px 14px;list-style:none;-webkit-box-align:center;align-items:center}.f-discussion-dialog-content .f-discussion-section-list-empty{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;-webkit-box-align:center;align-items:center;-webkit-box-pack:center;justify-content:center;height:100%}.f-discussion-dialog-content .f-discussion-section-list-empty .f-icon{margin-bottom:8px;font-size:26px;color:#e2e6f0}.f-discussion-dialog-content .f-discussion-section-list-empty .section-list-empty-text{font-size:13px;color:rgba(0,0,0,.25)}.f-discussion-dialog-content .f-discussion-personnel-item-tip{width:36px;height:36px;font-size:12px;color:#fff;text-align:center;line-height:32px;border-radius:50%;background-color:#4796ff}.f-discussion-dialog-content .f-discussion-section-list .section-list-item{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;-webkit-box-pack:justify;justify-content:space-between;height:32px;padding:0 12px;cursor:pointer}.f-discussion-dialog-content .f-discussion-section-list .section-list-item-name{font-size:14px;color:rgba(0,0,0,.85);line-height:32px}.f-discussion-dialog-content .f-discussion-section-list .section-list-item-next{flex-shrink:0;position:relative;padding-left:20px;padding-right:10px;line-height:20px;font-size:12px;color:#388fff;cursor:pointer}.f-discussion-dialog-content .f-discussion-section-list .section-list-item-next::before{position:absolute;left:0;top:50%;width:1px;height:12px;margin-top:-6px;background:#d9d9d9;content:''}.f-discussion-dialog-content .f-discussion-section-list .section-list-item:hover{background:#f3f8ff}.f-discussion-dialog-content .f-discussion-section-list .section-list-item .section-list-item-checkbox.custom-control.custom-checkbox{margin-right:0;padding:0;margin-bottom:0}.f-discussion-dialog-content .f-discussion-section-list .person-list-item{height:auto;padding-top:7px;padding-bottom:6px}.f-discussion-dialog-content .f-discussion-section-list .person-list-item-img{flex-shrink:0;width:36px;height:36px;border-radius:50%;overflow:hidden}.f-discussion-dialog-content .f-discussion-section-list .person-list-item-img img{display:block;width:100%;height:100%}.f-discussion-dialog-content .f-discussion-section-list .person-list-item-content{padding-left:6px}.f-discussion-dialog-content .f-discussion-section-list .person-list-item-content .section-list-item-name{font-size:14px;color:rgba(0,0,0,.85);line-height:20px}.f-discussion-dialog-content .f-discussion-section-list .person-list-item-content .section-list-item-tip{font-size:12px;color:rgba(0,0,0,.35);line-height:20px}.f-discussion-person-list-page{padding:10px 0;text-align:center;cursor:pointer}.f-discussion-person-list-page .section-page-text{font-size:14px;color:rgba(0,0,0,.55)}.f-discussion-person-list-page:hover .section-page-text{color:rgba(0,0,0,.85)}"]
                }] }
    ];
    /** @nocollapse */
    DiscussionEditorComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef },
        { type: ChangeDetectorRef },
        { type: NotifyService },
        { type: Injector },
        { type: HttpClient }
    ]; };
    DiscussionEditorComponent.propDecorators = {
        personnelsPrimaryKey: [{ type: Input }],
        personnelsDisplayKey: [{ type: Input }],
        replyPersonnelsDisplayKey: [{ type: Input }],
        cancelVisible: [{ type: Input }],
        editHeight: [{ type: Input }],
        type: [{ type: Input }],
        orgUrl: [{ type: Input }],
        personSearchUrl: [{ type: Input }],
        treeIdField: [{ type: Input }],
        treeDisplayText: [{ type: Input }],
        sectionData: [{ type: Input }],
        personnels: [{ type: Input }],
        replyUser: [{ type: Input }],
        attachFiles: [{ type: Input }],
        selectionsChange: [{ type: Output }],
        lineDataChange: [{ type: Output }],
        valueChange: [{ type: Output }],
        filePreview: [{ type: Output }],
        fileRemove: [{ type: Output }],
        fileUploadDone: [{ type: Output }],
        personnelSearch: [{ type: Output }],
        getOugUsers: [{ type: Output }],
        editor: [{ type: ViewChild, args: ['editor',] }],
        personModal: [{ type: ViewChild, args: ['personModal',] }]
    };
    return DiscussionEditorComponent;
}());
export { DiscussionEditorComponent };
if (false) {
    /** @type {?} */
    DiscussionEditorComponent.prototype.personnelsPrimaryKey;
    /** @type {?} */
    DiscussionEditorComponent.prototype.personnelsDisplayKey;
    /** @type {?} */
    DiscussionEditorComponent.prototype.replyPersonnelsDisplayKey;
    /** @type {?} */
    DiscussionEditorComponent.prototype.cancelVisible;
    /** @type {?} */
    DiscussionEditorComponent.prototype.editHeight;
    /** @type {?} */
    DiscussionEditorComponent.prototype.type;
    /** @type {?} */
    DiscussionEditorComponent.prototype.orgUrl;
    /** @type {?} */
    DiscussionEditorComponent.prototype.personSearchUrl;
    /** @type {?} */
    DiscussionEditorComponent.prototype.treeIdField;
    /** @type {?} */
    DiscussionEditorComponent.prototype.treeDisplayText;
    /** @type {?} */
    DiscussionEditorComponent.prototype.sectionData;
    /** @type {?} */
    DiscussionEditorComponent.prototype.options;
    /** @type {?} */
    DiscussionEditorComponent.prototype.placeholder;
    /** @type {?} */
    DiscussionEditorComponent.prototype.sectiontreeData;
    /**
     * @type {?}
     * @private
     */
    DiscussionEditorComponent.prototype._personnels;
    /**
     * @type {?}
     * @private
     */
    DiscussionEditorComponent.prototype._replyUser;
    /** @type {?} */
    DiscussionEditorComponent.prototype._attachFiles;
    /** @type {?} */
    DiscussionEditorComponent.prototype.selectionsChange;
    /** @type {?} */
    DiscussionEditorComponent.prototype.lineDataChange;
    /** @type {?} */
    DiscussionEditorComponent.prototype.valueChange;
    /**
     * 附件预览
     * @type {?}
     */
    DiscussionEditorComponent.prototype.filePreview;
    /**
     * 附件删除
     * @type {?}
     */
    DiscussionEditorComponent.prototype.fileRemove;
    /** @type {?} */
    DiscussionEditorComponent.prototype.fileUploadDone;
    /** @type {?} */
    DiscussionEditorComponent.prototype.personnelSearch;
    /** @type {?} */
    DiscussionEditorComponent.prototype.getOugUsers;
    /** @type {?} */
    DiscussionEditorComponent.prototype.editor;
    /** @type {?} */
    DiscussionEditorComponent.prototype.personModal;
    /** @type {?} */
    DiscussionEditorComponent.prototype.textValue;
    /** @type {?} */
    DiscussionEditorComponent.prototype.tempTextValue;
    /** @type {?} */
    DiscussionEditorComponent.prototype.tempPersonnelsValue;
    /** @type {?} */
    DiscussionEditorComponent.prototype.tempSectionValue;
    /** @type {?} */
    DiscussionEditorComponent.prototype.personModalVisible;
    /** @type {?} */
    DiscussionEditorComponent.prototype.selectedPersonnels;
    /** @type {?} */
    DiscussionEditorComponent.prototype.selectedSection;
    /** @type {?} */
    DiscussionEditorComponent.prototype.personnelText;
    /** @type {?} */
    DiscussionEditorComponent.prototype.copyPersonnels;
    /** @type {?} */
    DiscussionEditorComponent.prototype.innerPersonnels;
    /** @type {?} */
    DiscussionEditorComponent.prototype.el;
    /** @type {?} */
    DiscussionEditorComponent.prototype.token;
    /** @type {?} */
    DiscussionEditorComponent.prototype.relativeVisible;
    /** @type {?} */
    DiscussionEditorComponent.prototype.permission;
    /** @type {?} */
    DiscussionEditorComponent.prototype.attachFilesModalVisible;
    /** @type {?} */
    DiscussionEditorComponent.prototype.pageSize;
    /** @type {?} */
    DiscussionEditorComponent.prototype.searchPersonnelList;
    /** @type {?} */
    DiscussionEditorComponent.prototype.showSearchList;
    /** @type {?} */
    DiscussionEditorComponent.prototype.permissionList;
    /** @type {?} */
    DiscussionEditorComponent.prototype.localeService;
    /** @type {?} */
    DiscussionEditorComponent.prototype.groupIcon;
    /**
     * @type {?}
     * @private
     */
    DiscussionEditorComponent.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    DiscussionEditorComponent.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    DiscussionEditorComponent.prototype.cd;
    /**
     * @type {?}
     * @private
     */
    DiscussionEditorComponent.prototype.notifyService;
    /**
     * @type {?}
     * @private
     */
    DiscussionEditorComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    DiscussionEditorComponent.prototype.http;
}
