// @edp-bif/common-api/esm5/lib/slim-entity/impl/AbstractRtSlimBillProcessEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/slim-entity/impl/AbstractRtSlimBillProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from "../../ref";
import { AbstractSlimBillProcessEntity } from "./AbstractSlimBillProcessEntity";
/**
 * 运行时简化单据流实体类。
 * @abstract
 */
var /**
 * 运行时简化单据流实体类。
 * @abstract
 */
AbstractRtSlimBillProcessEntity = /** @class */ (function (_super) {
    tslib_1.__extends(AbstractRtSlimBillProcessEntity, _super);
    function AbstractRtSlimBillProcessEntity() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isEnabled = false;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    AbstractRtSlimBillProcessEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        this.isEnabled = false;
        this.eventRouterConditionsId = null;
        this.mappingConditionsId = null;
        this.mappingModeId = null;
        this.saveModeId = null;
    };
    /**
     * @return {?}
     */
    AbstractRtSlimBillProcessEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsId", this.eventRouterConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsId", this.mappingConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeId", this.mappingModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeId", this.saveModeId);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AbstractRtSlimBillProcessEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.eventRouterConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsId");
        this.mappingConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsId");
        this.mappingModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeId");
        this.saveModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeId");
    };
    return AbstractRtSlimBillProcessEntity;
}(AbstractSlimBillProcessEntity));
/**
 * 运行时简化单据流实体类。
 * @abstract
 */
export { AbstractRtSlimBillProcessEntity };
if (false) {
    /** @type {?} */
    AbstractRtSlimBillProcessEntity.prototype.isEnabled;
    /** @type {?} */
    AbstractRtSlimBillProcessEntity.prototype.eventRouterConditionsId;
    /** @type {?} */
    AbstractRtSlimBillProcessEntity.prototype.mappingConditionsId;
    /** @type {?} */
    AbstractRtSlimBillProcessEntity.prototype.mappingModeId;
    /** @type {?} */
    AbstractRtSlimBillProcessEntity.prototype.saveModeId;
}
