// @edp-bif/common-api/esm5/lib/slim-entity/impl/RtSlimBillCreationProcessEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/slim-entity/impl/RtSlimBillCreationProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from "../../ref";
import { AbstractRtSlimBillProcessEntity } from "./AbstractRtSlimBillProcessEntity";
/**
 * 运行时简化生单流实体类。
 */
var /**
 * 运行时简化生单流实体类。
 */
RtSlimBillCreationProcessEntity = /** @class */ (function (_super) {
    tslib_1.__extends(RtSlimBillCreationProcessEntity, _super);
    function RtSlimBillCreationProcessEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    RtSlimBillCreationProcessEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.formSettingsId = null;
        this.creationRuleId = null;
        this.downBillVoProxyMetadataId = null;
        this.downBillVoProxyId = null;
        this.upBillTrackingSettingsId = null;
        this.downBillTrackingSettingsId = null;
        this.reverseWriteRuleId = null;
        this.taskCenterConfig4Push = null;
        this.taskCenterConfig4Pull = null;
        this.isSupportPull = false;
        this.isSupportPush = false;
    };
    /**
     * @return {?}
     */
    RtSlimBillCreationProcessEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "formSettingsId", this.formSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataId", this.downBillVoProxyMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyId", this.downBillVoProxyId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettingsId", this.upBillTrackingSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettingsId", this.downBillTrackingSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleId", this.reverseWriteRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Push", this.taskCenterConfig4Push);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Pull", this.taskCenterConfig4Pull);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPull", this.isSupportPull);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPush", this.isSupportPush);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RtSlimBillCreationProcessEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.formSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettingsId");
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.downBillVoProxyMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataId");
        this.downBillVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyId");
        this.upBillTrackingSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettingsId");
        this.downBillTrackingSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettingsId");
        this.reverseWriteRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleId");
        this.taskCenterConfig4Push = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Push");
        this.taskCenterConfig4Pull = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Pull");
        this.isSupportPull = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPull");
        this.isSupportPush = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPush");
    };
    return RtSlimBillCreationProcessEntity;
}(AbstractRtSlimBillProcessEntity));
/**
 * 运行时简化生单流实体类。
 */
export { RtSlimBillCreationProcessEntity };
if (false) {
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.isSupportPush;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.formSettingsId;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.creationRuleId;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.downBillVoProxyMetadataId;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.downBillVoProxyId;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.upBillTrackingSettingsId;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.downBillTrackingSettingsId;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.reverseWriteRuleId;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.taskCenterConfig4Push;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.taskCenterConfig4Pull;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.isSupportPull;
}
