// @farris/extend-scroll-collapsible-area/esm5/lib/scroll-area.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/scroll-area.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef } from '@angular/core';
import { fromEvent } from 'rxjs';
import { debounceTime } from 'rxjs/operators';
var FEScrollAreaDirective = /** @class */ (function () {
    function FEScrollAreaDirective(el) {
        this.el = el;
        // 记录滚动的值
        this.scrollAreaTop = 0;
        // 记录值
        this.threshold = 0;
    }
    /**
     * @return {?}
     */
    FEScrollAreaDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.bindScrollEvent();
    };
    /**
     * @return {?}
     */
    FEScrollAreaDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.scrollSub.unsubscribe();
        this.scrollSub = null;
    };
    // 追加关联
    // 追加关联
    /**
     * @param {?} cInstance
     * @return {?}
     */
    FEScrollAreaDirective.prototype.addCollapseInstance = 
    // 追加关联
    /**
     * @param {?} cInstance
     * @return {?}
     */
    function (cInstance) {
        this.relatedCollapseInstance = cInstance;
    };
    // 绑定事件
    // 绑定事件
    /**
     * @return {?}
     */
    FEScrollAreaDirective.prototype.bindScrollEvent = 
    // 绑定事件
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.scrollSub = fromEvent(this.el.nativeElement, 'scroll').pipe(debounceTime(20)).subscribe((/**
         * @return {?}
         */
        function () {
            //判断滚动的方向
            _this.scrollDirection();
        }));
    };
    /**
     * @return {?}
     */
    FEScrollAreaDirective.prototype.scrollDirection = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var afterScrollTop = this.el.nativeElement.scrollTop;
        /** @type {?} */
        var delta = afterScrollTop - this.scrollAreaTop;
        // 没有变化
        if (delta === 0)
            return false;
        this.scrollAreaTop = afterScrollTop;
        this.judgeThreshold(delta > 0 ? "down" : "up");
    };
    /**
     * @private
     * @param {?} direction
     * @return {?}
     */
    FEScrollAreaDirective.prototype.judgeThreshold = /**
     * @private
     * @param {?} direction
     * @return {?}
     */
    function (direction) {
        this.relatedCollapseInstance.updateHeight(this.scrollAreaTop, direction);
    };
    FEScrollAreaDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[fe-scroll-area]'
                },] }
    ];
    /** @nocollapse */
    FEScrollAreaDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    return FEScrollAreaDirective;
}());
export { FEScrollAreaDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FEScrollAreaDirective.prototype.scrollSub;
    /**
     * @type {?}
     * @private
     */
    FEScrollAreaDirective.prototype.relatedCollapseInstance;
    /**
     * @type {?}
     * @private
     */
    FEScrollAreaDirective.prototype.scrollAreaTop;
    /**
     * @type {?}
     * @private
     */
    FEScrollAreaDirective.prototype.threshold;
    /**
     * @type {?}
     * @private
     */
    FEScrollAreaDirective.prototype.el;
}
