// @gsp-wf/wf-process-editor/esm5/lib/views/save-as-edit/save-as-edit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter, ViewChild, TemplateRef, Injector, Optional, ChangeDetectorRef, Inject, LOCALE_ID } from '@angular/core';
import { of } from 'rxjs';
import { LookupGridComponent, ServerSideToken } from '@farris/ui-lookup';
import { HttpService } from '@ecp-caf/caf-common';
import { switchMap } from 'rxjs/operators';
import { LookupDataService } from '@farris/lookup-config';
import { MessagerService } from '@farris/ui-messager';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../i18n/pipe/wf-process-editor-locale.pipe';
var SaveAsEditComponent = /** @class */ (function () {
    function SaveAsEditComponent(injector, http, message, cdr, localeId, resources, localePipe) {
        var _this = this;
        this.injector = injector;
        this.http = http;
        this.message = message;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.submit = new EventEmitter();
        this.cancel = new EventEmitter();
        this.accessLevel = 'public';
        this.isAdministratorConfig = false;
        this.enableAdminOrgAut = false;
        this.dataUri = '/api/runtime/sys/v1.0/sysOrgs';
        this.loadType = 'layerload';
        this.disabled = false;
        // private _organizations: any[];
        // set organizations(data: any[]) {
        //   this._organizations = data;
        //   if (this._organizations && this._organizations.length > 0) {
        //     this.items = this.makeTreeByParent(this._organizations, 'parentId', 'id');
        //   } else {
        //     this.items = new Array();
        //   }
        // }
        // get organizations() {
        //   return this._organizations;
        // }
        this.items = new Array();
        this.accessLevelCollection = [
            { value: 'public', text: this.localePipe.transform('component.saveAsEdit.public') },
            { value: 'private', text: this.localePipe.transform('component.saveAsEdit.private') }
        ];
        this.columns = [
            { title: this.localePipe.transform('component.saveAsEdit.code'), field: 'code', width: '100', searchField: 'code' },
            { title: this.localePipe.transform('component.saveAsEdit.name'), field: 'name', width: '100', searchField: 'name' },
        ];
        this.dictPicking = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.lookUp.http = {
                getData: (/**
                 * @param {?} uri
                 * @param {?} params
                 * @return {?}
                 */
                function (uri, params) {
                    console.log(params);
                    /** @type {?} */
                    var searchValue = JSON.parse(params.searchValue);
                    /** @type {?} */
                    var parent = searchValue.parentId;
                    /** @type {?} */
                    var param;
                    if (_this.lookUp._searchState && _this.lookUp._searchState.field && _this.lookUp._searchState.value) {
                        if (_this.lookUp._searchState.field === '*') {
                            param = { codeOrName: _this.lookUp._searchState.value };
                        }
                        else if (_this.lookUp._searchState.field === 'name') {
                            param = { name: _this.lookUp._searchState.value };
                        }
                        else {
                            param = { code: _this.lookUp._searchState.value };
                        }
                    }
                    /** @type {?} */
                    var filter;
                    if (_this.isAdministratorConfig) {
                        filter = _this.administrator.administratorAuthorities.map((/**
                         * @param {?} a
                         * @return {?}
                         */
                        function (a) { return ({
                            id: a.organization, rule: a.includeSubordinate ? 'SelectedAndChild' : 'Selected'
                        }); }));
                    }
                    return _this.getSysOrgs(parent, param, filter).pipe(switchMap((/**
                     * @param {?} re
                     * @return {?}
                     */
                    function (re) {
                        console.log(re);
                        return of((/** @type {?} */ ({
                            columns: _this.columns,
                            displayInfo: { idField: 'id', valueField: 'id', textField: 'name' },
                            displayType: 'TreeList',
                            idField: 'id',
                            valueField: 'id',
                            textField: 'name',
                            searchFields: [
                                { label: 'code', value: 'code' }, { label: 'name', value: 'name' }
                            ],
                            title: _this.localePipe.transform('component.saveAsEdit.organization'),
                            total: 0,
                            treeInfo: {
                                dataField: 'treeinfo',
                                isDetailField: 'isDetail',
                                layerField: 'layer',
                                layerType: 'parentId',
                                loadDataType: 'async',
                                onlySelectLeaf: false,
                                parentField: 'parentElement',
                                pathField: 'path',
                                treeDataIsInit: true
                            },
                            items: _this.orgTreeData(re, param)
                        })));
                    })));
                })
            };
            return of(true);
        });
        this.dictPicked = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.organizationName = data.name;
            _this.organization = data.id;
            return of(true);
        });
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
    }
    /**
     * @return {?}
     */
    SaveAsEditComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    SaveAsEditComponent.prototype.saveAsOK = /**
     * @return {?}
     */
    function () {
        this.disabled = true;
        if (this.validator()) {
            this.submit.next({
                procName: this.procName,
                accessLevel: this.accessLevel,
                organization: this.organization,
                organizationName: this.organizationName
            });
        }
    };
    /**
     * @return {?}
     */
    SaveAsEditComponent.prototype.saveAsCancel = /**
     * @return {?}
     */
    function () {
        this.cancel.emit();
    };
    /**
     * @return {?}
     */
    SaveAsEditComponent.prototype.validator = /**
     * @return {?}
     */
    function () {
        if (!this.procName) {
            this.message.warning(this.localePipe.transform('component.saveAsEdit.nameNotEmpty'));
            this.disabled = false;
            return false;
        }
        if (!this.organization) {
            this.message.warning(this.localePipe.transform('component.saveAsEdit.organizationNotEmpty'));
            this.disabled = false;
            return false;
        }
        return true;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    SaveAsEditComponent.prototype.search = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        /** @type {?} */
        var tt = (/** @type {?} */ (this.lookUp.componentRef.instance));
        tt.searchHandle.search(data.field, data.value, 'client');
    };
    /**
     * @param {?} data
     * @return {?}
     */
    SaveAsEditComponent.prototype.selectedDataHandler = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.organization = data.id;
        this.organizationName = data.name;
    };
    /**
     * @private
     * @param {?} orgs
     * @param {?} searchValue
     * @return {?}
     */
    SaveAsEditComponent.prototype.orgTreeData = /**
     * @private
     * @param {?} orgs
     * @param {?} searchValue
     * @return {?}
     */
    function (orgs, searchValue) {
        var e_1, _a;
        /** @type {?} */
        var result = new Array();
        /** @type {?} */
        var treeData = new Array();
        try {
            for (var orgs_1 = tslib_1.__values(orgs), orgs_1_1 = orgs_1.next(); !orgs_1_1.done; orgs_1_1 = orgs_1.next()) {
                var org = orgs_1_1.value;
                /** @type {?} */
                var entity = {};
                entity.data = org;
                entity.children = [];
                entity.id = org.id;
                entity.leaf = org.isDetail;
                entity.selectable = org.farris_selectable;
                treeData.push(entity);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (orgs_1_1 && !orgs_1_1.done && (_a = orgs_1.return)) _a.call(orgs_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        /** @type {?} */
        var obj = {};
        // 使用对象重新存储数据
        /** @type {?} */
        var len = treeData.length;
        for (var i = 0; i < len; i++) {
            /** @type {?} */
            var d = treeData[i];
            obj[treeData[i].id] = treeData[i];
        }
        // 遍历数据
        for (var j = 0; j < len; j++) {
            /** @type {?} */
            var data = treeData[j];
            /** @type {?} */
            var parentData = obj[data.data.parentId];
            if (parentData) {
                if (!parentData.children) {
                    parentData.children = new Array();
                }
                parentData.children.push(data);
            }
            else {
                result.push(data);
            }
        }
        // 过滤的情况下，不包含下级的节点设置为叶子节点
        if (searchValue) {
            this.treeIterator(result);
        }
        return result;
    };
    /**
     * @private
     * @param {?} treeData
     * @return {?}
     */
    SaveAsEditComponent.prototype.treeIterator = /**
     * @private
     * @param {?} treeData
     * @return {?}
     */
    function (treeData) {
        var e_2, _a;
        if (treeData && treeData.length > 0) {
            try {
                for (var treeData_1 = tslib_1.__values(treeData), treeData_1_1 = treeData_1.next(); !treeData_1_1.done; treeData_1_1 = treeData_1.next()) {
                    var d = treeData_1_1.value;
                    if (d.children && d.children.length > 0) {
                        this.treeIterator(d.children);
                    }
                    else {
                        d.leaf = true;
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (treeData_1_1 && !treeData_1_1.done && (_a = treeData_1.return)) _a.call(treeData_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
    };
    /**
     * @private
     * @param {?} item
     * @param {?} orgs
     * @return {?}
     */
    SaveAsEditComponent.prototype.findChild = /**
     * @private
     * @param {?} item
     * @param {?} orgs
     * @return {?}
     */
    function (item, orgs) {
        var _this = this;
        /** @type {?} */
        var childList = orgs.filter((/**
         * @param {?} childitem
         * @return {?}
         */
        function (childitem) {
            if (childitem.parentId === item.id) {
                return true;
            }
        }));
        if (childList && childList.length > 0) {
            childList.forEach((/**
             * @param {?} child
             * @return {?}
             */
            function (child) {
                /** @type {?} */
                var childEntity = {};
                childEntity.data = child;
                childEntity.children = [];
                childEntity.id = child.id;
                childEntity.leaf = child.isDetail;
                childEntity.selectable = child.farris_selectable;
                item.children.push(childEntity);
                _this.findChild(childEntity, orgs);
            }));
        }
    };
    /** 按照parent码构造树 */
    /**
     * 按照parent码构造树
     * @private
     * @param {?} datas
     * @param {?} parentField
     * @param {?} keyField
     * @return {?}
     */
    SaveAsEditComponent.prototype.makeTreeByParent = /**
     * 按照parent码构造树
     * @private
     * @param {?} datas
     * @param {?} parentField
     * @param {?} keyField
     * @return {?}
     */
    function (datas, parentField, keyField) {
        var _this = this;
        // 上級爲空，或者沒有上級的
        /** @type {?} */
        var pathLayerOnes = datas.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            if (!x[parentField]) {
                return true;
            }
            /** @type {?} */
            var p = datas.find((/**
             * @param {?} parent
             * @return {?}
             */
            function (parent) { return parent[keyField] === x[parentField]; }));
            if (!p) {
                return true;
            }
        }));
        /** @type {?} */
        var treedata = [];
        pathLayerOnes.forEach((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            /** @type {?} */
            var entity = {};
            entity.data = x;
            entity.children = [];
            treedata.push(entity);
            // this.root = x['id'];
        }));
        treedata.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            _this.findChildrenByParent(item, datas, parentField, keyField);
        }));
        return treedata;
    };
    /** 找到子节点 */
    /**
     * 找到子节点
     * @private
     * @param {?} parent
     * @param {?} datas
     * @param {?} parentField
     * @param {?} keyField
     * @return {?}
     */
    SaveAsEditComponent.prototype.findChildrenByParent = /**
     * 找到子节点
     * @private
     * @param {?} parent
     * @param {?} datas
     * @param {?} parentField
     * @param {?} keyField
     * @return {?}
     */
    function (parent, datas, parentField, keyField) {
        var _this = this;
        /** @type {?} */
        var childList = datas.filter((/**
         * @param {?} childitem
         * @return {?}
         */
        function (childitem) {
            if (childitem[parentField] === parent.data[keyField]) {
                return true;
            }
        }));
        if (childList && childList.length > 0) {
            childList.forEach((/**
             * @param {?} child
             * @return {?}
             */
            function (child) {
                /** @type {?} */
                var childEntity = {};
                childEntity.data = child;
                childEntity.children = [];
                parent.children.push(childEntity);
                _this.findChildrenByParent(childEntity, datas, parentField, keyField);
            }));
        }
    };
    /**
     * @private
     * @param {?} parent
     * @param {?} searchValue
     * @param {?} filter
     * @return {?}
     */
    SaveAsEditComponent.prototype.getSysOrgs = /**
     * @private
     * @param {?} parent
     * @param {?} searchValue
     * @param {?} filter
     * @return {?}
     */
    function (parent, searchValue, filter) {
        /** @type {?} */
        var url = "/api/runtime/sys/v1.0/sysOrgs";
        /** @type {?} */
        var param;
        if (searchValue) {
            param = Object.assign(searchValue, {
                IsSearchTree: true,
                withPermission: true,
                ruleFilter: filter
            });
        }
        else {
            if (parent) {
                param = Object.assign({}, {
                    parentId: parent,
                    withPermission: true,
                    ruleFilter: filter
                });
            }
            else {
                param = Object.assign({}, {
                    parentId: '',
                    layer: 1,
                    withPermission: true,
                    ruleFilter: filter
                });
            }
        }
        url += "?param=" + JSON.stringify(param);
        return this.http.get(url);
    };
    SaveAsEditComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-save-as-edit',
                    template: "<!-- <div class=\"f-page\">\r\n  <div class=\"f-page-main\">\r\n    <div class=\"d-flex m-2 py-2 farris-row\">\r\n      <div class=\"w-100 px-4\">\r\n        <label class=\"col-form-label\">\r\n          <span class=\"farris-label-text\">\u6D41\u7A0B\u540D\u79F0</span>\r\n        </label>\r\n        <div class=\"farris-input-wrap\">\r\n          <input type=\"text\" class=\"form-control\" [(ngModel)]=\"procName\" name=\"procName\" />\r\n          <div class=\"farris-feedback valid-feedback\"></div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n    <div class=\"d-flex m-2 py-2 farris-row\">\r\n      <div class=\"px-4 w-100\">\r\n        <label class=\"col-form-label\">\r\n          <span class=\"farris-label-text\">\u6D41\u7A0BKey\u503C</span>\r\n        </label>\r\n        <div class=\"farris-input-wrap\">\r\n          <input type=\"text\" class=\"form-control\" [(ngModel)]=\"procKey\" name=\"procKey\" [readonly]=\"true\" />\r\n          <div class=\"farris-feedback valid-feedback\"></div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</div> -->\r\n\r\n<div class=\"f-page f-page-card f-page-is-mainsubcard\">\r\n  <section class=\"f-page-main\">\r\n    <!--\u4E0D\u786E\u8BA4---------------------------\u5361\u7247\u7EC4\u4EF6\u4E0D\u540C\u7684Section\u662F\u5728\u4E0D\u540C\u7684\u7EC4\u4EF6\u91CC\uFF0C\u8FD8\u662F\u5728\u540C\u4E00\u4E2A\u5361\u7247\u7EC4\u4EF6\u91CC-->\r\n    <!--\u8868\u5355\u533A\u57DF-->\r\n    <!--\u9700\u8981\u5D4C\u5957\u7684\u5C42---------------------------IDE\u8F6C\u5316\u6210\u7EC4\u4EF6----------->\r\n    <div class=\"f-struct-wrapper\">\r\n      <farris-section class=\"f-section-form f-section-in-mainsubcard\" [enableAccordion]=\"''\" mainTitle=\"{{'component.saveAsEdit.basicAttributes'|wfLocale}}\"\r\n        [subTitle]=\"''\" [enableMaximize]=\"false\">\r\n\r\n        <!--\u8F93\u5165\u63A7\u4EF6 \u6CA1\u6709\u5206\u7EC4-->\r\n        <div class=\"f-form-layout farris-form farris-form-controls-inline pr-4 m-2\">\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-info text-danger\">*</span>\r\n                  <span class=\"farris-label-text\">{{'component.saveAsEdit.processName' | wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <input type=\"text\" class=\"form-control\" [(ngModel)]=\"procName\" name=\"procName\" />\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{'component.saveAsEdit.processKeyValue' | wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <input type=\"text\" class=\"form-control\" [(ngModel)]=\"procKey\" name=\"procKey\" [readonly]=\"true\" />\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-info text-danger\">*</span>\r\n                  <span class=\"farris-label-text\">{{'component.saveAsEdit.maintenanceOrg' | wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <farris-lookup-grid *ngIf=\"!enableAdminOrgAut; else adminOrg\" #lookup [(ngModel)]=\"organizationName\" name=\"orgName\"\r\n                    [singleSelect]=\"true\" [dictPicked]=\"dictPicked\"\r\n                    [enableClear]=\"false\"\r\n                    [editable]=\"false\"\r\n                    [dictPicking]=\"dictPicking\"\r\n                    [uri]=\"dataUri\"\r\n                    [loadTreeDataType]=\"loadType\">\r\n                  </farris-lookup-grid>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-info text-danger\">*</span>\r\n                  <span class=\"farris-label-text\">{{'component.saveAsEdit.accessLevel' | wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <farris-combo-list [(ngModel)]=\"accessLevel\" name=\"accessLevel\"\r\n                    [ngModelOptions]=\"{standalone: true}\" [idField]=\"'value'\" [textField]=\"'text'\"\r\n                    [valueField]=\"'value'\" [data]=\"accessLevelCollection\" [panelHeight]=\"'auto'\" [enableClear]=\"false\">\r\n                  </farris-combo-list>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </farris-section>\r\n    </div>\r\n  </section>\r\n</div>\r\n\r\n<ng-template #buttonRef>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm mx-2\" (click)=\"saveAsCancel()\">{{'component.saveAsEdit.saveAsCancel' | wfLocale}}</button>\r\n  <button type=\"button\" class=\"btn btn-sm btn-primary mx-2\" [disabled]=\"disabled\" (click)=\"saveAsOK()\">{{'component.saveAsEdit.saveAsOK' | wfLocale}}</button>\r\n</ng-template>\r\n\r\n\r\n<ng-template #adminOrg>\r\n  <farris-lookup-grid #adminOrgLookup [(ngModel)]=\"organizationName\" name=\"adminOrgName\"\r\n    [helpId]=\"'processdesigneradminorgdataaut'\" [enableClear]=\"false\" [editable]=\"false\"\r\n    [singleSelect]=\"true\" [uri]=\"'assets/data/tree-list-help.json'\" (selectedData)=\"selectedDataHandler($event)\">\r\n  </farris-lookup-grid>\r\n</ng-template>\r\n",
                    providers: [
                        { provide: ServerSideToken, useClass: LookupDataService },
                        WfLocalePipe,
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    SaveAsEditComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: HttpService },
        { type: MessagerService },
        { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
        { type: WfLocalePipe, decorators: [{ type: Optional }] }
    ]; };
    SaveAsEditComponent.propDecorators = {
        btnRef: [{ type: ViewChild, args: ['buttonRef',] }],
        procName: [{ type: Input }],
        procKey: [{ type: Input }],
        lookUp: [{ type: ViewChild, args: ['lookup',] }],
        submit: [{ type: Output }],
        cancel: [{ type: Output }]
    };
    return SaveAsEditComponent;
}());
export { SaveAsEditComponent };
if (false) {
    /** @type {?} */
    SaveAsEditComponent.prototype.btnRef;
    /** @type {?} */
    SaveAsEditComponent.prototype.procName;
    /** @type {?} */
    SaveAsEditComponent.prototype.procKey;
    /** @type {?} */
    SaveAsEditComponent.prototype.lookUp;
    /** @type {?} */
    SaveAsEditComponent.prototype.submit;
    /** @type {?} */
    SaveAsEditComponent.prototype.cancel;
    /** @type {?} */
    SaveAsEditComponent.prototype.accessLevel;
    /** @type {?} */
    SaveAsEditComponent.prototype.organization;
    /** @type {?} */
    SaveAsEditComponent.prototype.organizationName;
    /** @type {?} */
    SaveAsEditComponent.prototype.isAdministratorConfig;
    /** @type {?} */
    SaveAsEditComponent.prototype.enableAdminOrgAut;
    /** @type {?} */
    SaveAsEditComponent.prototype.administrator;
    /** @type {?} */
    SaveAsEditComponent.prototype.dataUri;
    /** @type {?} */
    SaveAsEditComponent.prototype.loadType;
    /** @type {?} */
    SaveAsEditComponent.prototype.disabled;
    /** @type {?} */
    SaveAsEditComponent.prototype.items;
    /** @type {?} */
    SaveAsEditComponent.prototype.accessLevelCollection;
    /** @type {?} */
    SaveAsEditComponent.prototype.columns;
    /** @type {?} */
    SaveAsEditComponent.prototype.dictPicking;
    /** @type {?} */
    SaveAsEditComponent.prototype.dictPicked;
    /**
     * @type {?}
     * @private
     */
    SaveAsEditComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    SaveAsEditComponent.prototype.http;
    /**
     * @type {?}
     * @private
     */
    SaveAsEditComponent.prototype.message;
    /**
     * @type {?}
     * @private
     */
    SaveAsEditComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    SaveAsEditComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    SaveAsEditComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    SaveAsEditComponent.prototype.localePipe;
}
