// @edp-bif/common-api/fesm2015/edp-bif-common-api.js
import { JsonConvertableImpl, BpmnModelJsonHelper, Entity, TaskCenterContent, TaskHandlingMenuArgs, AifLogEntity, BizEventEntity, BizActionEntity, CreationRuleEntity, CreationProcessEntity, WritebackRuleEntity, WritebackProcessEntity, GeneralTextEntity, SlimEntity, BillSchemaField } from '@edp-aif/common-api';
export { JsonConvertableImpl, BpmnModelJsonHelper, TextValue, SlimEntity, Entity, CallType, GeneralTextEntity, FormSettings } from '@edp-aif/common-api';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/ref/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/ConditionType.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const ConditionType = {
    /**
     * 增量表达式。
     */
    // INCREMENTAL_EXPRESSION = "INCREMENTAL_EXPRESSION",
    /**
     * 全量表达式。
     */
    // FULL_EXPRESSION = "FULL_EXPRESSION",
    /**
     * 表达式。
     */
    EXPRESSION: "EXPRESSION",
    /**
     * 构件。
     */
    COMPONENT: "COMPONENT",
    /**
    * VO动作。
    */
    VOACTION: "VOACTION",
    /**
     * 实现类。
     */
    IMPLCLASS: "IMPLCLASS",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ConditionEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 条件实体类。
 */
class ConditionEntity extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.isEnabled = false;
        this.remarks = null;
        this.supportAutoProcess = true;
        this.supportPushProcess = false;
        this.supportPullProcess = false;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.id = null;
        this.isEnabled = false;
        this.name = null;
        this.type = null;
        this.content = null;
        this.remarks = null;
        this.supportAutoProcess = true;
        this.supportPushProcess = false;
        this.supportPullProcess = false;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "id", this.id);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "type", this.type);
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        BpmnModelJsonHelper.SetAttributeValue(obj, "remarks", this.remarks);
        BpmnModelJsonHelper.SetAttributeValue(obj, "supportAutoProcess", this.supportAutoProcess);
        BpmnModelJsonHelper.SetAttributeValue(obj, "supportPushProcess", this.supportPushProcess);
        BpmnModelJsonHelper.SetAttributeValue(obj, "supportPullProcess", this.supportPullProcess);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.id = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "id");
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.type = ConditionType[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "type")];
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
        this.remarks = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "remarks");
        this.supportAutoProcess = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "supportAutoProcess");
        this.supportPushProcess = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "supportPushProcess");
        this.supportPullProcess = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "supportPullProcess");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/MappingMode.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const MappingMode = {
    /**
     * 映射规则。
     */
    MAPPINGRULE: "MAPPINGRULE",
    /**
     * 自定义映射构件。
     */
    COMPONENT: "COMPONENT",
    /**
     * VO动作。
     */
    VOACTION: "VOACTION",
    /**
     * 映射规则ID。
     */
    MAPPINGRULEID: "MAPPINGRULEID",
    /**
     * 表达式
     */
    EXPRESSION: "EXPRESSION",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/MappingModeEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 映射方式实体类。
 */
class MappingModeEntity extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.mappingMode = null;
        this.content = null;
        this.mappingRuleContent = null;
        this.componentContent = null;
        this.voActionContent = null;
        this.expressionContent = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingMode", this.mappingMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleContent", this.mappingRuleContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentContent", this.componentContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voActionContent", this.voActionContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "expressionContent", this.expressionContent);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.mappingMode = MappingMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingMode")];
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
        this.mappingRuleContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleContent");
        this.componentContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentContent");
        this.voActionContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voActionContent");
        this.expressionContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "expressionContent");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/SaveMode.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const SaveMode = {
    /**
     * 默认。
     */
    DEFAULT: "DEFAULT",
    /**
     * 构件。
     */
    COMPONENT: "COMPONENT",
    /**
     * Vo动作。
     */
    VOACTION: "VOACTION",
    /**
     * 任务中心
     */
    TASKCENTER: "TASKCENTER",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/SaveModeEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 保存模式实体类。
 */
class SaveModeEntity extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.saveMode = null;
        this.content = null;
        this.componentContent = null;
        this.voActionContent = null;
        this.taskCenterContent = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveMode", this.saveMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentContent", this.componentContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voActionContent", this.voActionContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterContent", this.taskCenterContent);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.saveMode = SaveMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveMode")];
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
        this.componentContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentContent");
        this.voActionContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voActionContent");
        this.taskCenterContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterContent");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/content/impl/AbstractRtBillProcessContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 单据流内容类。
 * @param TEntity 单据流实体。
 * @abstract
 * @template TEntity
 */
class AbstractRtBillProcessContent extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.entity = null;
        this.eventRouterConditionsContent = new Array();
        this.mappingConditionsContent = new Array();
        this.mappingModeContent = null;
        this.saveModeContent = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "entity", this.entity);
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsContent", this.eventRouterConditionsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsContent", this.mappingConditionsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeContent", this.mappingModeContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeContent", this.saveModeContent);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.entity = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "entity", this.getEntityClass());
        this.eventRouterConditionsContent = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "eventRouterConditionsContent", ConditionEntity);
        this.mappingConditionsContent = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingConditionsContent", ConditionEntity);
        this.mappingModeContent = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "mappingModeContent", MappingModeEntity);
        this.saveModeContent = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "saveModeContent", SaveModeEntity);
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/AbstractBillProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 单据流实体类。
 */
class AbstractBillProcessEntity extends Entity {
    constructor() {
        super(...arguments);
        this.enabledParentEventRouterCond = true;
        this.enabledParentMappingCond = true;
        this.sysInit = false;
        this.origin = true;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.code = null;
        this.name = null;
        this.description = null;
        this.persistenceVersion = null;
        this.version = 0;
        this.callType = null;
        // this.bizTypeL1 = null;
        // this.bizTypeL2 = null;
        // this.bizTypeL3 = null;
        // this.bizTypeL4 = null;
        this.upBillAppCode = null;
        this.upBillSuCode = null;
        this.upBillBoId = null;
        this.upBillBeMetadataId = null;
        this.upBillBeId = null;
        this.upBillVoMetadataId = null;
        this.upBillVoId = null;
        this.downBillAppCode = null;
        this.downBillSuCode = null;
        this.downBillBoId = null;
        this.downBillBeMetadataId = null;
        this.downBillBeId = null;
        this.downBillVoMetadataId = null;
        this.downBillVoId = null;
        this.isSupportAutomatinFlow = false;
        this.bizEventId = null;
        this.bizActionId = null;
        this.originId = null;
        this.rootId = null;
        this.parentId = null;
        this.enabledParentEventRouterCond = true;
        this.enabledParentMappingCond = true;
        this.idPath = null;
        this.versionTag = null;
        this.upBillBeMetadataRootId = null;
        this.upBillBeMetadataParentId = null;
        this.upBillBeMetadataIdPath = null;
        this.upBillBeRootId = null;
        this.upBillBeParentId = null;
        this.upBillBeIdPath = null;
        this.upBillVoMetadataRootId = null;
        this.upBillVoMetadataParentId = null;
        this.upBillVoMetadataIdPath = null;
        this.upBillVoRootId = null;
        this.upBillVoParentId = null;
        this.upBillVoIdPath = null;
        this.downBillBeMetadataRootId = null;
        this.downBillBeMetadataParentId = null;
        this.downBillBeMetadataIdPath = null;
        this.downBillBeRootId = null;
        this.downBillBeParentId = null;
        this.downBillBeIdPath = null;
        this.downBillVoMetadataRootId = null;
        this.downBillVoMetadataParentId = null;
        this.downBillVoMetadataIdPath = null;
        this.downBillVoRootId = null;
        this.downBillVoParentId = null;
        this.downBillVoIdPath = null;
        this.sysInit = false;
        this.origin = true;
        this.colI18nValues = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "description", this.description);
        BpmnModelJsonHelper.SetAttributeValue(obj, "persistenceVersion", this.persistenceVersion);
        BpmnModelJsonHelper.SetAttributeValue(obj, "version", this.version);
        BpmnModelJsonHelper.SetAttributeValue(obj, "callType", this.callType);
        // BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeL1", this.bizTypeL1);
        // BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeL2", this.bizTypeL2);
        // BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeL3", this.bizTypeL3);
        // BpmnModelJsonHelper.SetAttributeValue(obj, "bizTypeL4", this.bizTypeL4);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillAppCode", this.upBillAppCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillSuCode", this.upBillSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBoId", this.upBillBoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeMetadataId", this.upBillBeMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeId", this.upBillBeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataId", this.upBillVoMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoId", this.upBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillAppCode", this.downBillAppCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillSuCode", this.downBillSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBoId", this.downBillBoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeMetadataId", this.downBillBeMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeId", this.downBillBeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataId", this.downBillVoMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoId", this.downBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportAutomatinFlow", this.isSupportAutomatinFlow);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventId", this.bizEventId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionId", this.bizActionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originId", this.originId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "rootId", this.rootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "parentId", this.parentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "enabledParentEventRouterCond", this.enabledParentEventRouterCond);
        BpmnModelJsonHelper.SetAttributeValue(obj, "enabledParentMappingCond", this.enabledParentMappingCond);
        BpmnModelJsonHelper.SetAttributeValue(obj, "idPath", this.idPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "versionTag", this.versionTag);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeMetadataRootId", this.upBillBeMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeMetadataParentId", this.upBillBeMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeMetadataIdPath", this.upBillBeMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeRootId", this.upBillBeRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeParentId", this.upBillBeParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeIdPath", this.upBillBeIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataRootId", this.upBillVoMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataParentId", this.upBillVoMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataIdPath", this.upBillVoMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoRootId", this.upBillVoRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoParentId", this.upBillVoParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoIdPath", this.upBillVoIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeMetadataRootId", this.downBillBeMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeMetadataParentId", this.downBillBeMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeMetadataIdPath", this.downBillBeMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeRootId", this.downBillBeRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeParentId", this.downBillBeParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeIdPath", this.downBillBeIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataRootId", this.downBillVoMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataParentId", this.downBillVoMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataIdPath", this.downBillVoMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoRootId", this.downBillVoRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoParentId", this.downBillVoParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoIdPath", this.downBillVoIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sysInit", this.sysInit);
        BpmnModelJsonHelper.SetAttributeValue(obj, "origin", this.origin);
        BpmnModelJsonHelper.SetAttributeValue(obj, "colI18nValues", this.colI18nValues);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.description = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "description");
        this.persistenceVersion = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "persistenceVersion");
        this.version = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "version");
        this.callType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "callType");
        // this.bizTypeL1 = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeL1");
        // this.bizTypeL2 = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeL2");
        // this.bizTypeL3 = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeL3");
        // this.bizTypeL4 = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizTypeL4");
        this.upBillAppCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillAppCode");
        this.upBillSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillSuCode");
        this.upBillBoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBoId");
        this.upBillBeMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeMetadataId");
        this.upBillBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeId");
        this.upBillVoMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataId");
        this.upBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoId");
        this.downBillAppCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillAppCode");
        this.downBillSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillSuCode");
        this.downBillBoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBoId");
        this.downBillBeMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeMetadataId");
        this.downBillBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeId");
        this.downBillVoMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataId");
        this.downBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoId");
        this.isSupportAutomatinFlow = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportAutomatinFlow");
        this.bizEventId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventId");
        this.bizActionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionId");
        this.originId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originId");
        this.rootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "rootId");
        this.parentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "parentId");
        this.enabledParentEventRouterCond = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "enabledParentEventRouterCond");
        this.enabledParentMappingCond = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "enabledParentMappingCond");
        this.idPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "idPath");
        this.versionTag = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "versionTag");
        this.upBillBeMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeMetadataRootId");
        this.upBillBeMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeMetadataParentId");
        this.upBillBeMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeMetadataIdPath");
        this.upBillBeRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeRootId");
        this.upBillBeParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeParentId");
        this.upBillBeIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeIdPath");
        this.upBillVoMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataRootId");
        this.upBillVoMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataParentId");
        this.upBillVoMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataIdPath");
        this.upBillVoRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoRootId");
        this.upBillVoParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoParentId");
        this.upBillVoIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoIdPath");
        this.downBillBeMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeMetadataRootId");
        this.downBillBeMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeMetadataParentId");
        this.downBillBeMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeMetadataIdPath");
        this.downBillBeRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeRootId");
        this.downBillBeParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeParentId");
        this.downBillBeIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeIdPath");
        this.downBillVoMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataRootId");
        this.downBillVoMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataParentId");
        this.downBillVoMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataIdPath");
        this.downBillVoRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoRootId");
        this.downBillVoParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoParentId");
        this.downBillVoIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoIdPath");
        this.sysInit = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "sysInit");
        this.origin = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "origin");
        this.colI18nValues = BpmnModelJsonHelper.GetAttributeValue(obj, "colI18nValues");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/AbstractBillCreationProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 单据生单流实体类。
 */
class AbstractBillCreationProcessEntity extends AbstractBillProcessEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.creationRuleId = null;
        this.creationProcessId = null;
        this.downBillVoProxyMetadataId = null;
        this.downBillVoProxyId = null;
        this.downBillVoProxyMetadataRootId = null;
        this.downBillVoProxyMetadataParId = null;
        this.downBillVoProxyMetadataIdPath = null;
        this.downBillVoProxyRootId = null;
        this.downBillVoProxyParentId = null;
        this.downBillVoProxyIdPath = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessId", this.creationProcessId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataId", this.downBillVoProxyMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyId", this.downBillVoProxyId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataRootId", this.downBillVoProxyMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataParId", this.downBillVoProxyMetadataParId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataIdPath", this.downBillVoProxyMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyRootId", this.downBillVoProxyRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyParentId", this.downBillVoProxyParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyIdPath", this.downBillVoProxyIdPath);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.creationProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessId");
        this.downBillVoProxyMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataId");
        this.downBillVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyId");
        this.downBillVoProxyMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataRootId");
        this.downBillVoProxyMetadataParId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataParId");
        this.downBillVoProxyMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataIdPath");
        this.downBillVoProxyRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyRootId");
        this.downBillVoProxyParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyParentId");
        this.downBillVoProxyIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyIdPath");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/RtBillCreationProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class RtBillCreationProcessEntity extends AbstractBillCreationProcessEntity {
    constructor() {
        super(...arguments);
        this.isEnabled = false;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.isEnabled = false;
        this.eventRouterConditionsId = null;
        this.mappingConditionsId = null;
        this.mappingModeId = null;
        this.saveModeId = null;
        this.formSettingsId = null;
        this.dtBillCreationProcessId = null;
        this.upBillTrackingSettingsId = null;
        this.downBillTrackingSettingsId = null;
        this.reverseWriteRuleId = null;
        this.taskCenterConfig4Push = null;
        this.taskCenterConfig4Pull = null;
        this.isSupportPull = false;
        this.isSupportPush = false;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsId", this.eventRouterConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsId", this.mappingConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeId", this.mappingModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeId", this.saveModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "formSettingsId", this.formSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dtBillCreationProcessId", this.dtBillCreationProcessId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettingsId", this.upBillTrackingSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettingsId", this.downBillTrackingSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleId", this.reverseWriteRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Push", this.taskCenterConfig4Push);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Pull", this.taskCenterConfig4Pull);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPull", this.isSupportPull);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPush", this.isSupportPush);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.eventRouterConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsId");
        this.mappingConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsId");
        this.mappingModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeId");
        this.saveModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeId");
        this.formSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettingsId");
        this.dtBillCreationProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dtBillCreationProcessId");
        this.upBillTrackingSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettingsId");
        this.downBillTrackingSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettingsId");
        this.reverseWriteRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleId");
        this.taskCenterConfig4Push = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Push");
        this.taskCenterConfig4Pull = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Pull");
        this.isSupportPull = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPull");
        this.isSupportPush = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPush");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/content/impl/RtBillCreationProcessContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 单据生单流内容类。
 */
class RtBillCreationProcessContent extends AbstractRtBillProcessContent {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.formSettings = null;
        this.upBillTrackingSettings = null;
        this.downBillTrackingSettings = null;
        this.reverseWriteRuleContent = null;
        this.taskCenterConfig4Push = null;
        this.taskCenterConfig4Pull = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "formSettings", this.formSettings);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettings", this.upBillTrackingSettings);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettings", this.downBillTrackingSettings);
        BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleContent", this.reverseWriteRuleContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Push", this.taskCenterConfig4Push);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Pull", this.taskCenterConfig4Pull);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.formSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettings");
        this.upBillTrackingSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettings");
        this.downBillTrackingSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettings");
        this.reverseWriteRuleContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleContent");
        this.taskCenterConfig4Push = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Push");
        this.taskCenterConfig4Pull = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Pull");
    }
    /**
     * @protected
     * @return {?}
     */
    getEntityClass() {
        return RtBillCreationProcessEntity;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/AbstractBillWritebackProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 单据回写流实体类。
 */
class AbstractBillWritebackProcessEntity extends AbstractBillProcessEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.writebackRuleId = null;
        this.writebackProcessId = null;
        this.upBillVoProxyMetadataId = null;
        this.upBillVoProxyId = null;
        this.upBillVoProxyMetadataRootId = null;
        this.upBillVoProxyMetadataParentId = null;
        this.upBillVoProxyMetadataIdPath = null;
        this.upBillVoProxyRootId = null;
        this.upBillVoProxyParentId = null;
        this.upBillVoProxyIdPath = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "writebackRuleId", this.writebackRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "writebackProcessId", this.writebackProcessId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyMetadataId", this.upBillVoProxyMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyId", this.upBillVoProxyId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyMetadataRootId", this.upBillVoProxyMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyMetadataParentId", this.upBillVoProxyMetadataParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyMetadataIdPath", this.upBillVoProxyMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyRootId", this.upBillVoProxyRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyParentId", this.upBillVoProxyParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoProxyIdPath", this.upBillVoProxyIdPath);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.writebackRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackRuleId");
        this.writebackProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackProcessId");
        this.upBillVoProxyMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyMetadataId");
        this.upBillVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyId");
        this.upBillVoProxyMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyMetadataRootId");
        this.upBillVoProxyMetadataParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyMetadataParentId");
        this.upBillVoProxyMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyMetadataIdPath");
        this.upBillVoProxyRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyRootId");
        this.upBillVoProxyParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyParentId");
        this.upBillVoProxyIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoProxyIdPath");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/RtBillWritebackProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class RtBillWritebackProcessEntity extends AbstractBillWritebackProcessEntity {
    constructor() {
        super(...arguments);
        this.isEnabled = false;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.isEnabled = false;
        this.eventRouterConditionsId = null;
        this.mappingConditionsId = null;
        this.mappingModeId = null;
        this.saveModeId = null;
        this.dtBillWritebackProcessId = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsId", this.eventRouterConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsId", this.mappingConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeId", this.mappingModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeId", this.saveModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dtBillWritebackProcessId", this.dtBillWritebackProcessId);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.eventRouterConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsId");
        this.mappingConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsId");
        this.mappingModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeId");
        this.saveModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeId");
        this.dtBillWritebackProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dtBillWritebackProcessId");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/content/impl/RtBillWritebackProcessContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 单据回写流内容类。
 */
class RtBillWritebackProcessContent extends AbstractRtBillProcessContent {
    /**
     * @protected
     * @return {?}
     */
    getEntityClass() {
        return RtBillWritebackProcessEntity;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/content/impl/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/content/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/BillTrackingDataShownMode.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const BillTrackingDataShownMode = {
    /**
    * 默认。
    */
    DEFAULT: "DEFAULT",
    /**
     * 自定义显示字段。
     */
    CUSTOMFIELD: "CUSTOMFIELD",
    /**
     * 自定义表单。
     */
    CUSTOMFORM: "CUSTOMFORM",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/BillTrackingNodeShownMode.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const BillTrackingNodeShownMode = {
    /**
    * 默认。
    */
    DEFAULT: "DEFAULT",
    /**
     * 自定义。
     */
    CUSTOM: "CUSTOM",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/JointQueryMode.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const JointQueryMode = {
    /**
     * 默认。
     */
    DEFAULT: "DEFAULT",
    /**
     * 通过App联查。
     */
    APP: "APP",
    /**
     * 通过URL联查。
     */
    URL: "URL",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/TaskCenterConfig.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class TaskCenterConfig extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.jointQueryMode = JointQueryMode.APP;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.taskCenterContent = null;
        this.jointQueryMode = JointQueryMode.APP;
        this.joitQueryMenu = null;
        this.jointQueryUrl = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterContent", this.taskCenterContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryMode", this.jointQueryMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "joitQueryMenu", this.joitQueryMenu);
        BpmnModelJsonHelper.SetAttributeValue(obj, "jointQueryUrl", this.jointQueryUrl);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.taskCenterContent = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "taskCenterContent", TaskCenterContent);
        this.jointQueryMode = JointQueryMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryMode")];
        this.joitQueryMenu = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "joitQueryMenu", TaskHandlingMenuArgs);
        this.jointQueryUrl = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "jointQueryUrl");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/TaskCenterConfig4Pull.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class TaskCenterConfig4Pull extends TaskCenterConfig {
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/TaskCenterConfig4Push.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class TaskCenterConfig4Push extends TaskCenterConfig {
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/AbstractBillCpAndWpEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class AbstractBillCpAndWpEntity extends Entity {
    constructor() {
        super(...arguments);
        this.sysInit = false;
        this.isCreationProcess = false;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.code = null;
        this.name = null;
        this.persistenceVersion = null;
        this.parentId = null;
        this.upBillVoId = null;
        this.downBillVoId = null;
        this.sysInit = false;
        this.isCreationProcess = false;
        this.originId = null;
        this.rootId = null;
        this.colI18nValues = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "persistenceVersion", this.persistenceVersion);
        BpmnModelJsonHelper.SetAttributeValue(obj, "parentId", this.parentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoId", this.upBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoId", this.downBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sysInit", this.sysInit);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isCreationProcess", this.isCreationProcess);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originId", this.originId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "rootId", this.rootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "colI18nValues", this.colI18nValues);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.persistenceVersion = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "persistenceVersion");
        this.parentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "parentId");
        this.upBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoId");
        this.downBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoId");
        this.sysInit = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "sysInit");
        this.isCreationProcess = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isCreationProcess");
        this.originId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originId");
        this.rootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "rootId");
        this.colI18nValues = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "colI18nValues");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/BillTrackingConfigEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class BillTrackingConfigEntity extends Entity {
    constructor() {
        super(...arguments);
        this.isEnabled = false;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.suCode = null;
        this.classTypeName = null;
        this.isEnabled = false;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "classTypeName", this.classTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.classTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "classTypeName");
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/BillTrackingSettings.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class BillTrackingSettings extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.billTrackingNodeShownMode = BillTrackingNodeShownMode.DEFAULT;
        this.billTrackingDataShownMode = BillTrackingDataShownMode.DEFAULT;
    }
    ;
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.billTrackingNodeShownMode = BillTrackingNodeShownMode.DEFAULT;
        this.billNodeName = null;
        this.billNodeShownStyle = null;
        this.summaryFieldsInfo = null;
        this.billTrackingDataShownMode = BillTrackingDataShownMode.DEFAULT;
        this.customFieldContent = null;
        this.customFormContent = null;
        this.version = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "billNodeName", this.billNodeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billTrackingNodeShownMode", this.billTrackingNodeShownMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billNodeShownStyle", this.billNodeShownStyle);
        BpmnModelJsonHelper.SetAttributeValue(obj, "summaryFieldsInfo", this.summaryFieldsInfo);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billTrackingDataShownMode", this.billTrackingDataShownMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "customFieldContent", this.customFieldContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "customFormContent", this.customFormContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "version", this.version);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.billNodeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billNodeName");
        this.billTrackingNodeShownMode = BillTrackingDataShownMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billTrackingNodeShownMode")];
        this.billNodeShownStyle = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billNodeShownStyle");
        this.summaryFieldsInfo = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "summaryFieldsInfo");
        this.billTrackingDataShownMode = BillTrackingDataShownMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billTrackingDataShownMode")];
        this.customFieldContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "customFieldContent");
        this.customFormContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "customFormContent");
        this.version = (/** @type {?} */ (BpmnModelJsonHelper.GetAttributeValueAsString(obj, "version")));
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ConditionContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ConditionContent extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.unSatisfiedInformation = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "unSatisfiedInformation", this.unSatisfiedInformation);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.unSatisfiedInformation = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "unSatisfiedInformation");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ComponentConditionContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ComponentConditionContent extends ConditionContent {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.componentMetadataId = null;
        this.componentId = null;
        this.name = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentMetadataId", this.componentMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentId", this.componentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.componentMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentMetadataId");
        this.componentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentId");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ComponentContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ComponentContent extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.componentMetadataId = null;
        this.componentId = null;
        this.name = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentMetadataId", this.componentMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentId", this.componentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.componentMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentMetadataId");
        this.componentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentId");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/DtBillCpAndWpEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DtBillCpAndWpEntity extends AbstractBillCpAndWpEntity {
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/DtBillCreationProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DtBillCreationProcessEntity extends AbstractBillCreationProcessEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.eventRouterConditionsContent = null;
        this.mappingConditionsContent = null;
        this.mappingModeContent = null;
        this.saveModeContent = null;
        this.formSettingsContent = null;
        this.upBillTrackingSettings = null;
        this.downBillTrackingSettings = null;
        this.reverseWriteRuleContent = null;
        this.taskCenterConfig4Push = null;
        this.taskCenterConfig4Pull = null;
        this.isSupportPull = false;
        this.isSupportPush = false;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsContent", this.eventRouterConditionsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsContent", this.mappingConditionsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeContent", this.mappingModeContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeContent", this.saveModeContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "formSettingsContent", this.formSettingsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettings", this.upBillTrackingSettings);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettings", this.downBillTrackingSettings);
        BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleContent", this.reverseWriteRuleContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Push", this.taskCenterConfig4Push);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Pull", this.taskCenterConfig4Pull);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPull", this.isSupportPull);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPush", this.isSupportPush);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.eventRouterConditionsContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsContent");
        this.mappingConditionsContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsContent");
        this.mappingModeContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeContent");
        this.saveModeContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeContent");
        this.formSettingsContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettingsContent");
        this.upBillTrackingSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettings");
        this.downBillTrackingSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettings");
        this.reverseWriteRuleContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleContent");
        this.taskCenterConfig4Push = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Push");
        this.taskCenterConfig4Pull = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Pull");
        this.isSupportPull = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPull");
        this.isSupportPush = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPush");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/DtBillWritebackProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DtBillWritebackProcessEntity extends AbstractBillWritebackProcessEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.eventRouterConditionsContent = null;
        this.mappingConditionsContent = null;
        this.mappingModeContent = null;
        this.saveModeContent = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsContent", this.eventRouterConditionsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsContent", this.mappingConditionsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeContent", this.mappingModeContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeContent", this.saveModeContent);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.eventRouterConditionsContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsContent");
        this.mappingConditionsContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsContent");
        this.mappingModeContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeContent");
        this.saveModeContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeContent");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ExportLogResult.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ExportLogResult extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.aifLogEntities = new Array();
        this.rtBillCreationProcessEntities = new Array();
        this.rtBillWritebackProcessEntities = new Array();
        this.bizEventEntities = new Array();
        this.bizActionEntities = new Array();
        this.creationRuleEntities = new Array();
        this.creationProcessEntities = new Array();
        this.writebackRuleEntities = new Array();
        this.writebackProcessEntities = new Array();
        this.generalTextEntities = new Array();
        this.aifLogEntitiesMap = {};
        this.rtBillCreationProcessEntitiesMap = {};
        this.rtBillWritebackProcessEntitiesMap = {};
        this.bizEventEntitiesMap = {};
        this.bizActionEntitiesMap = {};
        this.creationRuleEntitiesMap = {};
        this.creationProcessEntitiesMap = {};
        this.writebackRuleEntitiesMap = {};
        this.writebackProcessEntitiesMap = {};
        this.generalTextEntitiesMap = {};
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.aifLogEntities = [];
        this.rtBillCreationProcessEntities = [];
        this.rtBillWritebackProcessEntities = [];
        this.bizEventEntities = [];
        this.bizActionEntities = [];
        this.creationRuleEntities = [];
        this.creationProcessEntities = [];
        this.writebackRuleEntities = [];
        this.writebackProcessEntities = [];
        this.generalTextEntities = [];
        this.aifLogEntitiesMap = {};
        this.rtBillCreationProcessEntitiesMap = {};
        this.rtBillWritebackProcessEntitiesMap = {};
        this.bizEventEntitiesMap = {};
        this.bizActionEntitiesMap = {};
        this.creationRuleEntitiesMap = {};
        this.creationProcessEntitiesMap = {};
        this.writebackRuleEntitiesMap = {};
        this.writebackProcessEntitiesMap = {};
        this.generalTextEntitiesMap = {};
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "aifLogEntities", BpmnModelJsonHelper.ToJson(this.aifLogEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "rtBillCreationProcessEntities", BpmnModelJsonHelper.ToJson(this.rtBillCreationProcessEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "rtBillWritebackProcessEntities", BpmnModelJsonHelper.ToJson(this.rtBillWritebackProcessEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventEntities", BpmnModelJsonHelper.ToJson(this.bizEventEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionEntities", BpmnModelJsonHelper.ToJson(this.bizActionEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleEntities", BpmnModelJsonHelper.ToJson(this.creationRuleEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessEntities", BpmnModelJsonHelper.ToJson(this.creationProcessEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "writebackRuleEntities", BpmnModelJsonHelper.ToJson(this.writebackRuleEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "writebackProcessEntities", BpmnModelJsonHelper.ToJson(this.writebackProcessEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "generalTextEntities", BpmnModelJsonHelper.ToJson(this.generalTextEntities));
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.aifLogEntities = BpmnModelJsonHelper.FromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "aifLogEntities"), AifLogEntity);
        this.rtBillCreationProcessEntities = BpmnModelJsonHelper.FromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "rtBillCreationProcessEntities"), RtBillCreationProcessEntity);
        this.rtBillWritebackProcessEntities = BpmnModelJsonHelper.FromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "rtBillWritebackProcessEntities"), RtBillWritebackProcessEntity);
        this.bizEventEntities = BpmnModelJsonHelper.FromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventEntities"), BizEventEntity);
        this.bizActionEntities = BpmnModelJsonHelper.FromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionEntities"), BizActionEntity);
        this.creationRuleEntities = BpmnModelJsonHelper.FromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleEntities"), CreationRuleEntity);
        this.creationProcessEntities = BpmnModelJsonHelper.FromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessEntities"), CreationProcessEntity);
        this.writebackRuleEntities = BpmnModelJsonHelper.FromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackRuleEntities"), WritebackRuleEntity);
        this.writebackProcessEntities = BpmnModelJsonHelper.FromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackProcessEntities"), WritebackProcessEntity);
        this.generalTextEntities = BpmnModelJsonHelper.FromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "generalTextEntities"), GeneralTextEntity);
        for (let entity of this.aifLogEntities) {
            this.aifLogEntitiesMap[entity.id] = entity;
        }
        for (let entity of this.rtBillCreationProcessEntities) {
            this.rtBillCreationProcessEntitiesMap[entity.id] = entity;
        }
        for (let entity of this.rtBillWritebackProcessEntities) {
            this.rtBillWritebackProcessEntitiesMap[entity.id] = entity;
        }
        for (let entity of this.bizEventEntities) {
            this.bizEventEntitiesMap[entity.id] = entity;
        }
        for (let entity of this.bizActionEntities) {
            this.bizActionEntitiesMap[entity.id] = entity;
        }
        for (let entity of this.creationRuleEntities) {
            this.creationRuleEntitiesMap[entity.id] = entity;
        }
        for (let entity of this.creationProcessEntities) {
            this.creationProcessEntitiesMap[entity.id] = entity;
        }
        for (let entity of this.writebackRuleEntities) {
            this.writebackRuleEntitiesMap[entity.id] = entity;
        }
        for (let entity of this.writebackProcessEntities) {
            this.writebackProcessEntitiesMap[entity.id] = entity;
        }
        for (let entity of this.generalTextEntities) {
            this.generalTextEntitiesMap[entity.id] = entity;
        }
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ExpressionConditionContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ExpressionConditionContent extends ConditionContent {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.expressionContent = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "expressionContent", this.expressionContent);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.expressionContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "expressionContent");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ImplClassConditionContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ImplClassConditionContent extends ConditionContent {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.implClassTypeName = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "implClassTypeName", this.implClassTypeName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.implClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "implClassTypeName");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/MappingComponentContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class MappingComponentContent extends ComponentContent {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.mappingConfig = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConfig", this.mappingConfig);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.mappingConfig = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConfig");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/MappingRuleContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class MappingRuleContent extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.mappingRuleMetadataId = null;
        this.mappingRuleId = null;
        this.mappingContent = null;
        this.mappingConfig = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleMetadataId", this.mappingRuleMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleId", this.mappingRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingContent", this.mappingContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConfig", this.mappingConfig);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.mappingRuleMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleMetadataId");
        this.mappingRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleId");
        this.mappingContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingContent");
        this.mappingConfig = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConfig");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/RtBillCpAndWpEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class RtBillCpAndWpEntity extends AbstractBillCpAndWpEntity {
    constructor() {
        super(...arguments);
        this.isEnabled = false;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.isEnabled = false;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/VoActionConditionContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class VoActionConditionContent extends ConditionContent {
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/VoActionContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class VoActionContent extends JsonConvertableImpl {
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/AssociationType.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const AssociationType = {
    /**
     * 默认。
     */
    DEFAULT: "DEFAULT",
    /**
     * 生成关系。
     */
    GENERATE: "GENERATE",
    /**
     * 参照关系。
     */
    REFER: "REFER",
    /**
     * 核销关系。
     */
    VERIFICATION: "VERIFICATION",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/DocumentAssociationEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DocumentAssociationEntity extends Entity {
    constructor() {
        super(...arguments);
        this.associationType = AssociationType.DEFAULT;
        this.sequenceNum = 0;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.associationType = AssociationType.DEFAULT;
        this.sequenceNum = 0;
        this.originDocuSucode = null;
        this.originDocuVoId = null;
        this.originDocuTypeId = null;
        this.originDocuDataId = null;
        this.originDocuItemId = null;
        this.originDocuItemCode = null;
        this.originDocuDetailId = null;
        this.originDocuDataIdPath = null;
        this.upDocuSucode = null;
        this.upDocuVoId = null;
        this.upDocuTypeId = null;
        this.upDocuDataId = null;
        this.upDocuItemId = null;
        this.upDocuDetailId = null;
        this.upDocuDataIdPath = null;
        this.upDocuCode = null;
        this.upDocuDate = null;
        this.upDocuAbstract = null;
        this.downDocuSucode = null;
        this.downDocuVoId = null;
        this.downDocuTypeId = null;
        this.downDocuDataId = null;
        this.downDocuItemId = null;
        this.downDocuDetailId = null;
        this.downDocuDataIdPath = null;
        this.downDocuCode = null;
        this.downDocuDate = null;
        this.downDocuAbstract = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "associationType", this.associationType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sequenceNum", this.sequenceNum);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuSucode", this.originDocuSucode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuVoId", this.originDocuVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuTypeId", this.originDocuTypeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuDataId", this.originDocuDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuItemId", this.originDocuItemId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuItemCode", this.originDocuItemCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuDetailId", this.originDocuDetailId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "originDocuDataIdPath", this.originDocuDataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuSucode", this.upDocuSucode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuVoId", this.upDocuVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuTypeId", this.upDocuTypeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDataId", this.upDocuDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuItemId", this.upDocuItemId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDetailId", this.upDocuDetailId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDataIdPath", this.upDocuDataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuCode", this.upDocuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDate", this.upDocuDate);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuAbstract", this.upDocuAbstract);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuSucode", this.downDocuSucode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuVoId", this.downDocuVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuTypeId", this.downDocuTypeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDataId", this.downDocuDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuItemId", this.downDocuItemId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDetailId", this.downDocuDetailId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDataIdPath", this.downDocuDataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuCode", this.downDocuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDate", this.downDocuDate);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuAbstract", this.downDocuAbstract);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.associationType = AssociationType[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "associationType")];
        this.sequenceNum = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "sequenceNum");
        this.originDocuSucode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuSucode");
        this.originDocuVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuVoId");
        this.originDocuTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuTypeId");
        this.originDocuDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuDataId");
        this.originDocuItemId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuItemId");
        this.originDocuItemCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuItemCode");
        this.originDocuDetailId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuDetailId");
        this.originDocuDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "originDocuDataIdPath");
        this.upDocuSucode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuSucode");
        this.upDocuVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuVoId");
        this.upDocuTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuTypeId");
        this.upDocuDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDataId");
        this.upDocuItemId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuItemId");
        this.upDocuDetailId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDetailId");
        this.upDocuDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDataIdPath");
        this.upDocuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuCode");
        this.upDocuDate = BpmnModelJsonHelper.GetAttributeValueAsDateTime(obj, "upDocuDate");
        this.upDocuAbstract = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuAbstract");
        this.downDocuSucode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuSucode");
        this.downDocuVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuVoId");
        this.downDocuTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuTypeId");
        this.downDocuDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDataId");
        this.downDocuItemId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuItemId");
        this.downDocuDetailId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDetailId");
        this.downDocuDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDataIdPath");
        this.downDocuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuCode");
        this.downDocuDate = BpmnModelJsonHelper.GetAttributeValueAsDateTime(obj, "downDocuDate");
        this.downDocuAbstract = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuAbstract");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/AssociationFieldsType.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const AssociationFieldsType = {
    /**
     * 默认。
     */
    DEFAULT: "DEFAULT",
    /**
     * 主对主。
     */
    MAIN_MAIN: "MAIN_MAIN",
    /**
     * 主对从。
     */
    MAIN_ITEM: "MAIN_ITEM",
    /**
     * 主对从从。
     */
    MAIN_DETAIL: "MAIN_DETAIL",
    /**
     * 从对主。
     */
    ITEM_MAIN: "ITEM_MAIN",
    /**
     * 从对从。
     */
    ITEM_ITEM: "ITEM_ITEM",
    /**
     * 从对从从。
     */
    ITEM_DETAIL: "ITEM_DETAIL",
    /**
     * 从从对主。
     */
    DETAIL_MAIN: "DETAIL_MAIN",
    /**
     * 从从对从。
     */
    DETAIL_ITEM: "DETAIL_ITEM",
    /**
     * 从从对从从。
     */
    DETAIL_DETAIL: "DETAIL_DETAIL",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/FieldsAssociationEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class FieldsAssociationEntity extends Entity {
    constructor() {
        super(...arguments);
        this.sysInit = false;
        // upAssociationFieldsValueType: AssociationFieldsValueType = AssociationFieldsValueType.FIELD;
        // upDocuConstant: string;
        // downAssociationFieldsValueType: AssociationFieldsValueType = AssociationFieldsValueType.FIELD;
        // downDocuConstant: string;
        this.associationFieldsType = AssociationFieldsType.DEFAULT;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.code = null;
        this.name = null;
        this.sysInit = false;
        // this.docuRelationId = null;
        // this.upAssociationFieldsValueType = AssociationFieldsValueType.FIELD;
        // this.downAssociationFieldsValueType = AssociationFieldsValueType.FIELD;
        // this.upDocuConstant = null;
        // this.downDocuConstant = null;
        this.associationFieldsType = AssociationFieldsType.DEFAULT;
        this.upDocuMainCode = null;
        this.upDocuMainField = null;
        this.upDocuMainFieldName = null;
        this.downDocuMainCode = null;
        this.downDocuMainField = null;
        this.downDocuMainFieldName = null;
        this.upDocuItemCode = null;
        this.upDocuItemField = null;
        this.upDocuItemFieldName = null;
        this.downDocuItemCode = null;
        this.downDocuItemField = null;
        this.downDocuItemFieldName = null;
        this.upDocuDetailCode = null;
        this.upDocuDetailField = null;
        this.upDocuDetailFieldName = null;
        this.downDocuDetailCode = null;
        this.downDocuDetailField = null;
        this.downDocuDetailFieldName = null;
        this.mapUpMainField = null;
        this.mapUpMainFieldName = null;
        this.mapDownMainField = null;
        this.mapDownMainFieldName = null;
        this.mapUpItemCode = null;
        this.mapUpItemField = null;
        this.mapUpItemFieldName = null;
        this.mapDownItemCode = null;
        this.mapDownItemField = null;
        this.mapDownItemFieldName = null;
        this.mapUpDetailCode = null;
        this.mapUpDetailField = null;
        this.mapUpDetailFieldName = null;
        this.mapDownDetailCode = null;
        this.mapDownDetailField = null;
        this.mapDownDetailFieldName = null;
        this.colI18nValues = null;
        this.suCode = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sysInit", this.sysInit);
        BpmnModelJsonHelper.SetAttributeValue(obj, "docuRelationId", this.docuRelationId);
        // BpmnModelJsonHelper.SetAttributeValue(obj, "upAssociationFieldsValueType", this.upAssociationFieldsValueType);
        // BpmnModelJsonHelper.SetAttributeValue(obj, "downAssociationFieldsValueType", this.downAssociationFieldsValueType);
        // BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuConstant", this.upDocuConstant);
        // BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuConstant", this.downDocuConstant);
        BpmnModelJsonHelper.SetAttributeValue(obj, "associationFieldsType", this.associationFieldsType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuMainCode", this.upDocuMainCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuMainField", this.upDocuMainField);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuMainFieldName", this.upDocuMainFieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuMainCode", this.downDocuMainCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuMainField", this.downDocuMainField);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuMainFieldName", this.downDocuMainFieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuItemCode", this.upDocuItemCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuItemField", this.upDocuItemField);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuItemFieldName", this.upDocuItemFieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuItemCode", this.downDocuItemCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuItemField", this.downDocuItemField);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuItemFieldName", this.downDocuItemFieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDetailCode", this.upDocuDetailCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDetailField", this.upDocuDetailField);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upDocuDetailFieldName", this.upDocuDetailFieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDetailCode", this.downDocuDetailCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDetailField", this.downDocuDetailField);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downDocuDetailFieldName", this.downDocuDetailFieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpMainField", this.mapUpMainField);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpMainFieldName", this.mapUpMainFieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownMainField", this.mapDownMainField);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownMainFieldName", this.mapDownMainFieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpItemCode", this.mapUpItemCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpItemField", this.mapUpItemField);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpItemFieldName", this.mapUpItemFieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownItemCode", this.mapDownItemCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownItemField", this.mapDownItemField);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownItemFieldName", this.mapDownItemFieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpDetailCode", this.mapUpDetailCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpDetailField", this.mapUpDetailField);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapUpDetailFieldName", this.mapUpDetailFieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownDetailCode", this.mapDownDetailCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownDetailField", this.mapDownDetailField);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mapDownDetailFieldName", this.mapDownDetailFieldName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "colI18nValues", this.colI18nValues);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.sysInit = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "sysInit");
        this.docuRelationId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "docuRelationId");
        // const upAssociationFieldsValueType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upAssociationFieldsValueType");
        // if (upAssociationFieldsValueType != null && upAssociationFieldsValueType != "") {
        //   this.upAssociationFieldsValueType = AssociationFieldsValueType[upAssociationFieldsValueType];
        // }
        // const downAssociationFieldsValueType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downAssociationFieldsValueType");
        // if (downAssociationFieldsValueType != null && downAssociationFieldsValueType != "") {
        //   this.downAssociationFieldsValueType = AssociationFieldsValueType[downAssociationFieldsValueType];
        // }
        // this.upDocuConstant = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuConstant");
        // this.downDocuConstant = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuConstant");
        this.associationFieldsType = AssociationFieldsType[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "associationFieldsType")];
        this.upDocuMainCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuMainCode");
        this.upDocuMainField = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuMainField");
        this.upDocuMainFieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuMainFieldName");
        this.downDocuMainCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuMainCode");
        this.downDocuMainField = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuMainField");
        this.downDocuMainFieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuMainFieldName");
        this.upDocuItemCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuItemCode");
        this.upDocuItemField = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuItemField");
        this.upDocuItemFieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuItemFieldName");
        this.downDocuItemCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuItemCode");
        this.downDocuItemField = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuItemField");
        this.downDocuItemFieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuItemFieldName");
        this.upDocuDetailCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDetailCode");
        this.upDocuDetailField = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDetailField");
        this.upDocuDetailFieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upDocuDetailFieldName");
        this.downDocuDetailCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDetailCode");
        this.downDocuDetailField = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDetailField");
        this.downDocuDetailFieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downDocuDetailFieldName");
        this.mapUpMainField = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpMainField");
        this.mapUpMainFieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpMainFieldName");
        this.mapDownMainField = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownMainField");
        this.mapDownMainFieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownMainFieldName");
        this.mapUpItemCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpItemCode");
        this.mapUpItemField = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpItemField");
        this.mapUpItemFieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpItemFieldName");
        this.mapDownItemCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownItemCode");
        this.mapDownItemField = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownItemField");
        this.mapDownItemFieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownItemFieldName");
        this.mapUpDetailCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpDetailCode");
        this.mapUpDetailField = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpDetailField");
        this.mapUpDetailFieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapUpDetailFieldName");
        this.mapDownDetailCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownDetailCode");
        this.mapDownDetailField = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownDetailField");
        this.mapDownDetailFieldName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mapDownDetailFieldName");
        this.colI18nValues = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "colI18nValues");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/AssociationLocation.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const AssociationLocation = {
    /**
     * 默认。
     */
    DEFAULT: "DEFAULT",
    /**
     * 上游单据记录。
     */
    UpBill: "UpBill",
    /**
     * 下游单据记录。
     */
    DownBill: "DownBill",
    /**
     * 中间表记录。
     */
    ThirdTable: "ThirdTable",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/RelationType.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const RelationType = {
    /**
     * 默认。
     */
    DEFAULT: "DEFAULT",
    /**
     * 一对多。
     */
    ONETOONE: "ONETOONE",
    /**
     * 一对多。
     */
    ONETOMANY: "ONETOMANY",
    /**
     * 多对一。
     */
    MANYTOONE: "MORETOONE",
    /**
     * 多对多。
     */
    MANYTOMANY: "MANYTOMANY",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/TrackingMode.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const TrackingMode = {
    /**
     * 默认。
     */
    DEFAULT: "DEFAULT",
    /**
     * 单据流方式。
     */
    AIF: "AIF",
    /**
     * 设计时配置表方式。
     */
    DTRELATION: "DTRELATION",
    /**
     * 运行时业务数据表方式。
     */
    RTRELATION: "RTRELATION",
    /**
     * 自定义构件方式。
     */
    COMPONENT: "COMPONENT",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/TrackingType.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const TrackingType = {
    /**
     * 默认。
     */
    DEFAULT: "DEFAULT",
    /**
     * 生成关系追踪。
     */
    GENERATE: "GENERATE",
    /**
     * 业务自定义。
     */
    BIZCUSTOM: "BIZCUSTOM",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/DocumentRelationEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DocumentRelationEntity extends Entity {
    constructor() {
        super(...arguments);
        this.sysInit = false;
        this.trackingType = TrackingType.DEFAULT;
        this.trackingMode = TrackingMode.DEFAULT;
        this.associationLocation = AssociationLocation.DEFAULT;
        this.relationType = RelationType.DEFAULT;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.code = null;
        this.name = null;
        this.sysInit = false;
        this.upDocuTypeId = null;
        this.downDocuTypeId = null;
        this.trackingType = TrackingType.DEFAULT;
        this.trackingMode = TrackingMode.DEFAULT;
        this.associationLocation = AssociationLocation.DEFAULT;
        this.associationSuCode = null;
        this.associationSchemaId = null;
        this.associationDataSourceType = null;
        this.relationType = RelationType.DEFAULT;
        this.colI18nValues = null;
        this.trackingComponentId = null;
        this.upConditionExpressionText = null;
        this.downConditionExpressionText = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, 'code', this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'name', this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'sysInit', this.sysInit);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'upDocuTypeId', this.upDocuTypeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'downDocuTypeId', this.downDocuTypeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'trackingType', this.trackingType);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'trackingMode', this.trackingMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'associationLocation', this.associationLocation);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'associationSuCode', this.associationSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'associationSchemaId', this.associationSchemaId);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'associationDataSourceType', this.associationDataSourceType);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'relationType', this.relationType);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'colI18nValues', this.colI18nValues);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'trackingComponentId', this.trackingComponentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'upConditionExpressionText', this.upConditionExpressionText);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'downConditionExpressionText', this.downConditionExpressionText);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'code');
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'name');
        this.sysInit = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, 'sysInit');
        this.upDocuTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'upDocuTypeId');
        this.downDocuTypeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'downDocuTypeId');
        this.trackingType = TrackingType[BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'trackingType')];
        this.trackingMode = TrackingMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'trackingMode')];
        this.associationLocation = AssociationLocation[BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'associationLocation')];
        this.associationSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'associationSuCode');
        this.associationSchemaId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'associationSchemaId');
        this.associationDataSourceType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'associationDataSourceType');
        this.relationType = RelationType[BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'relationType')];
        this.colI18nValues = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'colI18nValues');
        this.trackingComponentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'trackingComponentId');
        this.upConditionExpressionText = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'upConditionExpressionText');
        this.downConditionExpressionText = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'downConditionExpressionText');
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/DocumentTypeEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DocumentTypeEntity extends Entity {
    constructor() {
        super(...arguments);
        this.sysInit = false;
        // accessMode: AccessMode = AccessMode.DEFAULT;
        this.jointQueryMode = JointQueryMode.DEFAULT;
        // dataSql: string;
        this.extend = false;
    }
    /**
     * @return {?}
     */
    DocumentTypeEntity() {
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.code = null;
        this.name = null;
        this.sysInit = false;
        this.moduleId = null;
        this.bizKind = null;
        this.bizType = null;
        this.suCode = null;
        this.schemaId = null;
        this.jointQueryMode = JointQueryMode.DEFAULT;
        this.jointQueryParameters = null;
        this.nodeName = null;
        this.customNodeName = null;
        this.nodeStyle = null;
        this.customNodeStyle = null;
        this.summaryFields = null;
        this.customSummaryFields = null;
        this.displayFields = null;
        this.customDisplayFields = null;
        this.extend = false;
        this.colI18nValues = null;
        this.rootId = null;
        this.parentId = null;
        this.enableConditionExpressionText = null;
        this.trackingSourceType = null;
        this.tableSchema = null;
        this.customTableSchema = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, 'code', this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'name', this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'sysInit', this.sysInit);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'moduleId', this.moduleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'bizKind', this.bizKind);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'bizType', this.bizType);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'suCode', this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'schemaId', this.schemaId);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'jointQueryMode', this.jointQueryMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'jointQueryParameters', this.jointQueryParameters);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'nodeName', this.nodeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'customNodeName', this.customNodeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'nodeStyle', this.nodeStyle);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'customNodeStyle', this.customNodeStyle);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'summaryFields', this.summaryFields);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'customSummaryFields', this.customSummaryFields);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'displayFields', this.displayFields);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'customDisplayFields', this.customDisplayFields);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'extend', this.extend);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'colI18nValues', this.colI18nValues);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'rootId', this.rootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'parentId', this.parentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'enableConditionExpressionText', this.enableConditionExpressionText);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'trackingSourceType', this.trackingSourceType);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'tableSchema', this.tableSchema);
        BpmnModelJsonHelper.SetAttributeValue(obj, 'customTableSchema', this.customTableSchema);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'code');
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'name');
        this.sysInit = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, 'sysInit');
        this.moduleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'moduleId');
        this.bizKind = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'bizKind');
        this.bizType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'bizType');
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'suCode');
        this.schemaId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'schemaId');
        this.jointQueryMode = JointQueryMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'jointQueryMode')];
        this.jointQueryParameters = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'jointQueryParameters');
        this.nodeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'nodeName');
        this.customNodeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'customNodeName');
        this.nodeStyle = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'nodeStyle');
        this.customNodeStyle = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'customNodeStyle');
        this.summaryFields = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'summaryFields');
        this.customSummaryFields = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'customSummaryFields');
        this.displayFields = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'displayFields');
        this.customDisplayFields = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'customDisplayFields');
        this.extend = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, 'extend');
        this.colI18nValues = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'colI18nValues');
        this.rootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'rootId');
        this.parentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'parentId');
        this.enableConditionExpressionText = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'enableConditionExpressionText');
        this.trackingSourceType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'trackingSourceType');
        this.tableSchema = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'tableSchema');
        this.customTableSchema = BpmnModelJsonHelper.GetAttributeValueAsString(obj, 'customTableSchema');
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ThirdSchemaInfoEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ThirdSchemaInfoEntity extends Entity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.schemaCategoryId = null;
        this.schemaCategoryName = null;
        this.schemaConvertorClassTypeName = null;
        this.trackingConfigClassTypeName = null;
        this.suCode = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaCategoryId", this.schemaCategoryId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaCategoryName", this.schemaCategoryName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaConvertorClassTypeName", this.schemaConvertorClassTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "trackingConfigClassTypeName", this.trackingConfigClassTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.schemaCategoryId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaCategoryId");
        this.schemaCategoryName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaCategoryName");
        this.schemaConvertorClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaConvertorClassTypeName");
        this.trackingConfigClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "trackingConfigClassTypeName");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/TableSchema.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class TableSchema extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.maxDisplayNumber = 2000;
        this.isMainTable = false;
        this.isDisplay = true;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.code = null;
        this.maxDisplayNumber = 2000;
        this.isMainTable = false;
        this.isDisplay = true;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "maxDisplayNumber", this.maxDisplayNumber);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isMainTable", this.isMainTable);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isDisplay", this.isDisplay);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.maxDisplayNumber = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "maxDisplayNumber");
        this.isMainTable = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isMainTable");
        this.isDisplay = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isDisplay");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/AssociationFieldsValueType.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const AssociationFieldsValueType = {
    /**
     * 常量方式
     */
    CONSTANT: "CONSTANT",
    /**
     * 字段方式
     */
    FIELD: "FIELD",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/slim-entity/impl/AbstractSlimBillProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 简化单据流实体类。
 * @abstract
 */
class AbstractSlimBillProcessEntity extends SlimEntity {
    constructor() {
        super(...arguments);
        this.sysInit = false;
        this.enabledParentEventRouterCond = true;
        this.enabledParentMappingCond = true;
        this.origin = true;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.code = null;
        this.upBillVoMetadataId = null;
        this.upBillVoId = null;
        this.downBillVoMetadataId = null;
        this.downBillVoId = null;
        this.bizEventId = null;
        this.isSupportAutomatinFlow = false;
        this.sysInit = false;
        this.parentId = null;
        this.enabledParentEventRouterCond = true;
        this.enabledParentMappingCond = true;
        this.upBillBeId = null;
        this.downBillBeId = null;
        this.origin = true;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoMetadataId", this.upBillVoMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoId", this.upBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoMetadataId", this.downBillVoMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoId", this.downBillVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventId", this.bizEventId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportAutomatinFlow", this.isSupportAutomatinFlow);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sysInit", this.sysInit);
        BpmnModelJsonHelper.SetAttributeValue(obj, "parentId", this.parentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "enabledParentEventRouterCond", this.enabledParentEventRouterCond);
        BpmnModelJsonHelper.SetAttributeValue(obj, "enabledParentMappingCond", this.enabledParentMappingCond);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillBeId", this.upBillBeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillBeId", this.downBillBeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "origin", this.origin);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.upBillVoMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoMetadataId");
        this.upBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoId");
        this.downBillVoMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoMetadataId");
        this.downBillVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoId");
        this.bizEventId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventId");
        this.isSupportAutomatinFlow = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportAutomatinFlow");
        this.sysInit = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "sysInit");
        this.parentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "parentId");
        this.enabledParentEventRouterCond = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "enabledParentEventRouterCond");
        this.enabledParentMappingCond = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "enabledParentMappingCond");
        this.upBillBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillBeId");
        this.downBillBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillBeId");
        this.origin = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "origin");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/slim-entity/impl/AbstractRtSlimBillProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 运行时简化单据流实体类。
 * @abstract
 */
class AbstractRtSlimBillProcessEntity extends AbstractSlimBillProcessEntity {
    constructor() {
        super(...arguments);
        this.isEnabled = false;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        this.isEnabled = false;
        this.eventRouterConditionsId = null;
        this.mappingConditionsId = null;
        this.mappingModeId = null;
        this.saveModeId = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsId", this.eventRouterConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsId", this.mappingConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeId", this.mappingModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeId", this.saveModeId);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.eventRouterConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsId");
        this.mappingConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsId");
        this.mappingModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeId");
        this.saveModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeId");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/slim-entity/impl/RtSlimBillCreationProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 运行时简化生单流实体类。
 */
class RtSlimBillCreationProcessEntity extends AbstractRtSlimBillProcessEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.formSettingsId = null;
        this.creationRuleId = null;
        this.downBillVoProxyMetadataId = null;
        this.downBillVoProxyId = null;
        this.upBillTrackingSettingsId = null;
        this.downBillTrackingSettingsId = null;
        this.reverseWriteRuleId = null;
        this.taskCenterConfig4Push = null;
        this.taskCenterConfig4Pull = null;
        this.isSupportPull = false;
        this.isSupportPush = false;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "formSettingsId", this.formSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataId", this.downBillVoProxyMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyId", this.downBillVoProxyId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettingsId", this.upBillTrackingSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettingsId", this.downBillTrackingSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "reverseWriteRuleId", this.reverseWriteRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Push", this.taskCenterConfig4Push);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterConfig4Pull", this.taskCenterConfig4Pull);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPull", this.isSupportPull);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPush", this.isSupportPush);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.formSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettingsId");
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.downBillVoProxyMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataId");
        this.downBillVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyId");
        this.upBillTrackingSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettingsId");
        this.downBillTrackingSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettingsId");
        this.reverseWriteRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "reverseWriteRuleId");
        this.taskCenterConfig4Push = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Push");
        this.taskCenterConfig4Pull = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterConfig4Pull");
        this.isSupportPull = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPull");
        this.isSupportPush = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPush");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/slim-entity/impl/RtSlimBillCreationProcessEntityEx.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 单据生单流简化实体扩展类。
 */
class RtSlimBillCreationProcessEntityEx extends RtSlimBillCreationProcessEntity {
    /**
     * @param {?=} entity
     * @return {?}
     */
    RtSlimBillCreationProcessEntityEx(entity) {
        if (entity != null) {
            this.id = entity.id;
            this.code = entity.code;
            this.upBillVoId = entity.upBillVoId;
            this.upBillVoMetadataId = entity.upBillVoMetadataId;
            this.downBillVoId = entity.downBillVoId;
            this.downBillVoMetadataId = entity.downBillVoMetadataId;
            this.downBillVoProxyMetadataId = entity.downBillVoProxyMetadataId;
            this.downBillVoProxyId = entity.downBillVoProxyId;
            this.eventRouterConditionsId = entity.eventRouterConditionsId;
            this.mappingConditionsId = entity.mappingConditionsId;
            this.formSettingsId = entity.formSettingsId;
            this.bizEventId = entity.bizEventId;
            this.creationRuleId = entity.creationRuleId;
            this.isEnabled = entity.isEnabled;
            this.isSupportAutomatinFlow = entity.isSupportAutomatinFlow;
            this.name = entity.name;
            this.description = entity.description;
        }
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.name = null;
        this.description = null;
        this.upBillVoName = null;
        this.downBillVoName = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "description", this.description);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoName", this.upBillVoName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoName", this.downBillVoName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.description = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "description");
        this.upBillVoName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoName");
        this.downBillVoName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoName");
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/slim-entity/impl/RtSlimBillWritebackProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 运行时简化单据回写流实体类。
 */
class RtSlimBillWritebackProcessEntity extends AbstractRtSlimBillProcessEntity {
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/slim-entity/impl/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/slim-entity/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/front-entity/BillTrackerFormContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class BillTrackerFormContent {
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/front-entity/BillTrackerNodeShownStyle.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class BillTrackerNodeShownStyle {
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/front-entity/BillTrackerSchema.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class BillTrackerSchema {
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/front-entity/BillTrackerSchemaFieldContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class BillTrackerSchemaFieldContent {
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/front-entity/BillTrackerSchemaTable.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const fullPathPrefix = "#";
class BillTrackerSchemaTable {
    /**
     * 自定义JSON序列化方法
     * @return {?}
     */
    toJSON() {
        /** @type {?} */
        const fields = this.fields && this.fields.map((/**
         * @param {?} originF
         * @return {?}
         */
        (originF) => {
            /** @type {?} */
            let sf = new BillSchemaField();
            // sf.fullPath = originF.fullPath;
            sf.fieldPath = originF.fieldPath;
            sf.fieldName = originF.fieldName;
            sf.caption = originF.caption;
            sf.displayName = originF.displayName;
            sf.order = originF.order;
            sf.i18nCols = originF.i18nCols;
            sf.formatInfo = originF.formatInfo;
            return sf;
        }));
        /** @type {?} */
        const obj = {
            code: this.code,
            caption: this.caption,
            displayName: this.displayName,
            order: this.order,
            show: this.show,
            fields: fields,
            i18nCols: this.i18nCols,
            maxDisplayNumber: this.maxDisplayNumber,
            parentTable: this.parentTable,
        };
        return obj;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/front-entity/JointQueryParameter.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class JointQueryParameter {
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/front-entity/JointQueryParameters.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class JointQueryParameters {
    constructor() {
        this.params = [];
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/front-entity/JointQueryParameterType.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const JointQueryParameterType = {
    Constant: "Constant",
    Field: "Field",
    Expression: "Expression",
};

/**
 * @fileoverview added by tsickle
 * Generated from: lib/front-entity/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: lib/index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: edp-bif-common-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { AbstractRtBillProcessContent, RtBillCreationProcessContent, RtBillWritebackProcessContent, BillTrackingDataShownMode, BillTrackingNodeShownMode, ConditionType, MappingMode, SaveMode, TaskCenterConfig, TaskCenterConfig4Pull, TaskCenterConfig4Push, AbstractBillCpAndWpEntity, AbstractBillCreationProcessEntity, AbstractBillProcessEntity, AbstractBillWritebackProcessEntity, BillTrackingConfigEntity, BillTrackingSettings, ComponentConditionContent, ComponentContent, ConditionContent, ConditionEntity, DtBillCpAndWpEntity, DtBillCreationProcessEntity, DtBillWritebackProcessEntity, ExportLogResult, ExpressionConditionContent, ImplClassConditionContent, MappingComponentContent, MappingModeEntity, MappingRuleContent, RtBillCpAndWpEntity, RtBillCreationProcessEntity, RtBillWritebackProcessEntity, SaveModeEntity, VoActionConditionContent, VoActionContent, AssociationType, DocumentAssociationEntity, FieldsAssociationEntity, DocumentRelationEntity, DocumentTypeEntity, AssociationFieldsType, JointQueryMode, RelationType, TrackingMode, TrackingType, AssociationLocation, ThirdSchemaInfoEntity, TableSchema, AssociationFieldsValueType, AbstractRtSlimBillProcessEntity, AbstractSlimBillProcessEntity, RtSlimBillCreationProcessEntity, RtSlimBillCreationProcessEntityEx, RtSlimBillWritebackProcessEntity, BillTrackerFormContent, BillTrackerNodeShownStyle, BillTrackerSchema, BillTrackerSchemaFieldContent, fullPathPrefix, BillTrackerSchemaTable, JointQueryParameter, JointQueryParameters, JointQueryParameterType };
