// @qdp/formular/esm2015/lib/service/entitydata.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/entitydata.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class EntityDataService {
    constructor() {
        this.isFinish = false;
        this.formularInfo = {};
    }
    /**
     * @param {?} stepTye
     * @return {?}
     */
    setStepType(stepTye) {
        this.stepType = stepTye;
    }
    /**
     * @return {?}
     */
    getStepType() {
        return this.stepType;
    }
    /**
     * @param {?} isFinish
     * @param {?} formularInfo
     * @return {?}
     */
    setFormularState(isFinish, formularInfo) {
        this.isFinish = isFinish;
        this.formularInfo = formularInfo;
    }
    /**
     * @return {?}
     */
    getFormularState() {
        return { isFinish: this.isFinish, formularInfo: this.formularInfo };
    }
}
EntityDataService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
EntityDataService.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    EntityDataService.prototype.isFinish;
    /**
     * @type {?}
     * @private
     */
    EntityDataService.prototype.formularInfo;
    /**
     * @type {?}
     * @private
     */
    EntityDataService.prototype.stepType;
}
