// @qdp/formular/esm2015/lib/service/formular.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/formular.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ComponentFactoryResolver, Injector, Inject } from '@angular/core';
import { Subject, of } from 'rxjs';
import { MessagerService } from '@farris/ui-messager';
import { BsModalService } from '@farris/ui-modal';
import { FormularDefineWizardComponent } from '../component/formular-define-wizard/formular-define-wizard.component';
import { EntityDataService } from './entitydata.service';
import { BE_SERVER_URI_TOKEN } from '@farris/bef';
import { LocalizeService } from '@qdp/localize';
export class FormularService {
    /**
     * @param {?} modalService
     * @param {?} injector
     * @param {?} cfr
     * @param {?} msgService
     * @param {?} entityDataService
     * @param {?} url
     * @param {?} localizeService
     */
    constructor(modalService, injector, cfr, msgService, entityDataService, url, localizeService) {
        this.modalService = modalService;
        this.injector = injector;
        this.cfr = cfr;
        this.msgService = msgService;
        this.entityDataService = entityDataService;
        this.url = url;
        this.localizeService = localizeService;
        this.isFinish = true;
    }
    /**
     * @param {?} currentSU
     * @param {?} currentFormularId
     * @param {?=} formularType
     * @return {?}
     */
    showFormularDefine(currentSU, currentFormularId, formularType) {
        /** @type {?} */
        const formularResult = new Subject();
        /** @type {?} */
        const cmpF = this.cfr.resolveComponentFactory(FormularDefineWizardComponent);
        /** @type {?} */
        const cmpR = cmpF.create(this.injector);
        cmpR.instance.currentSU = currentSU;
        cmpR.instance.currentFormularId = currentFormularId;
        cmpR.instance.serverHost = this.url ? this.url : 'http://localhost:5100';
        cmpR.instance.formularType = formularType;
        this.dlg = this.modalService.show(cmpR, {
            title: this.localizeService.getValue('formular.service.labelFormulaWizard'),
            width: 1000, height: 800,
            showButtons: false,
            beforeClose: (/**
             * @param {?} ref
             * @return {?}
             */
            (ref) => {
                /** @type {?} */
                const formularState = this.entityDataService.getFormularState();
                if (formularState.isFinish) {
                    formularResult.next(formularState.formularInfo);
                    return of(true);
                }
                else {
                    /** @type {?} */
                    const canClose = new Subject();
                    this.msgService.question(this.localizeService.getValue('formular.service.messageAlert.formulaVerify'), (/**
                     * @return {?}
                     */
                    () => {
                        formularResult.next({ id: '', name: '' });
                        canClose.next(true);
                    }), (/**
                     * @return {?}
                     */
                    () => {
                        canClose.next(false);
                    }));
                    return canClose.asObservable();
                }
            })
        });
        cmpR.instance.parentDialog = this.dlg;
        return formularResult;
    }
}
FormularService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FormularService.ctorParameters = () => [
    { type: BsModalService },
    { type: Injector },
    { type: ComponentFactoryResolver },
    { type: MessagerService },
    { type: EntityDataService },
    { type: undefined, decorators: [{ type: Inject, args: [BE_SERVER_URI_TOKEN,] }] },
    { type: LocalizeService }
];
if (false) {
    /** @type {?} */
    FormularService.prototype.dlg;
    /** @type {?} */
    FormularService.prototype.isFinish;
    /**
     * @type {?}
     * @private
     */
    FormularService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    FormularService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    FormularService.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    FormularService.prototype.msgService;
    /**
     * @type {?}
     * @private
     */
    FormularService.prototype.entityDataService;
    /**
     * @type {?}
     * @private
     */
    FormularService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    FormularService.prototype.localizeService;
}
