// @gsp-cmp/chgdr/esm5/lib/chgdr-row-baseinfo-form/chgdr-row-baseinfo-form.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/chgdr-row-baseinfo-form/chgdr-row-baseinfo-form.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Injector, Input } from '@angular/core';
import { ChgdrTreeNode } from '../chgdr-data-viewer/chgdr-tree-builder';
import { I18nUtil } from '../service/i18n-service';
var ChgdrRowBaseinfoFormComponent = /** @class */ (function () {
    function ChgdrRowBaseinfoFormComponent(injector) {
        this.injector = injector;
        this.showDataId = false;
    }
    /**
     * @return {?}
     */
    ChgdrRowBaseinfoFormComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    Object.defineProperty(ChgdrRowBaseinfoFormComponent.prototype, "technologyPrimaryKeyLabel", {
        get: /**
         * @return {?}
         */
        function () {
            return I18nUtil.instant("technology-primary-key-label");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ChgdrRowBaseinfoFormComponent.prototype, "changeUserLabel", {
        get: /**
         * @return {?}
         */
        function () {
            return I18nUtil.instant("change-user-label");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ChgdrRowBaseinfoFormComponent.prototype, "changeTimeLabel", {
        get: /**
         * @return {?}
         */
        function () {
            return I18nUtil.instant("change-time-label");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ChgdrRowBaseinfoFormComponent.prototype, "changeOperateType", {
        get: /**
         * @return {?}
         */
        function () {
            return I18nUtil.instant("change-operate-type");
        },
        enumerable: true,
        configurable: true
    });
    ChgdrRowBaseinfoFormComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-chgdr-row-baseinfo-form',
                    template: "<div *ngIf=\"!!chgdrTreeNode\" class=\"farris-form farris-form-controls-inline\">\r\n  <form class=\"f-section-formgroup\">\r\n    <div class=\"f-section-formgroup-inputs\">\r\n      <div class=\"col-12 row\">\r\n\r\n        <div class=\"farris-group-wrap col-6\" *ngIf=\"showDataId\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\">\r\n              <span class=\"farris-label-text\">{{technologyPrimaryKeyLabel}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input type=\"text\" class=\"form-control\" [ngModel]=\"chgdrTreeNode.chgdrRow.dataId\" name=\"dataId\"\r\n                [readonly]=\"true\">\r\n              <div class=\"farris-feedback valid-feedback\"></div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n\r\n        <div class=\"farris-group-wrap col-6\" *ngFor=\"let field of chgdrTreeNode.chgdrRow.dataCodes\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\">\r\n              <span class=\"farris-label-text\">{{field.fieldName}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input type=\"text\" class=\"form-control\" [ngModel]=\"field.fieldValue\" name=\"dataCode-{{field.fieldName}}\"\r\n                [readonly]=\"true\">\r\n              <div class=\"farris-feedback valid-feedback\"></div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n\r\n        <div *ngIf=\"chgdrTreeNode.isRoot\" class=\"farris-group-wrap col-6\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\">\r\n              <span class=\"farris-label-text\">{{changeUserLabel}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input type=\"text\" class=\"form-control\" [ngModel]=\"chgdrTreeNode.chgdrData.userName\" name=\"userName\"\r\n                [readonly]=\"true\">\r\n              <div class=\"farris-feedback valid-feedback\"></div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n\r\n        <div *ngIf=\"chgdrTreeNode.isRoot\" class=\"farris-group-wrap col-6\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\">\r\n              <span class=\"farris-label-text\">{{changeTimeLabel}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input type=\"text\" class=\"form-control\" value=\"{{chgdrTreeNode.chgdrData.changeTime | date:'yyyy-MM-dd HH:mm:ss'}}\" name=\"changeTime\"\r\n                [readonly]=\"true\">\r\n              <div class=\"farris-feedback valid-feedback\"></div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n\r\n        <div class=\"farris-group-wrap col-6\">\r\n          <div class=\"form-group farris-form-group\">\r\n            <label class=\"col-form-label\">\r\n              <span class=\"farris-label-text\">{{changeOperateType}}</span>\r\n            </label>\r\n            <div class=\"farris-input-wrap\">\r\n              <input type=\"text\" class=\"form-control\" [ngModel]=\"chgdrTreeNode.chgdrRow.operateType.name\" name=\"operateType\"\r\n              [readonly]=\"true\">\r\n              <div class=\"farris-feedback valid-feedback\"></div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n        \r\n      </div>\r\n    </div>\r\n  </form>\r\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ChgdrRowBaseinfoFormComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    ChgdrRowBaseinfoFormComponent.propDecorators = {
        chgdrTreeNode: [{ type: Input }],
        showDataId: [{ type: Input }]
    };
    return ChgdrRowBaseinfoFormComponent;
}());
export { ChgdrRowBaseinfoFormComponent };
if (false) {
    /** @type {?} */
    ChgdrRowBaseinfoFormComponent.prototype.chgdrTreeNode;
    /** @type {?} */
    ChgdrRowBaseinfoFormComponent.prototype.showDataId;
    /**
     * @type {?}
     * @private
     */
    ChgdrRowBaseinfoFormComponent.prototype.injector;
}
