// @gsp-dip/data-imp-exp/esm5/lib/custom-export-dialog/export-dialog.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, TemplateRef, ComponentFactoryResolver } from '@angular/core';
import { FrameContext } from '@farris/devkit';
import { NotifyService } from '@farris/ui-notify';
import { DataExportService } from '../service/DataIEService/data-export.service';
import { FileOperateService } from '../service/DataIEService/file-operate.service';
import { EntityListService } from '@farris/command-services';
import { ElementSelectComponent } from '../element-select/element-select.component';
import { CustomDataIEService } from '../service/DataIEService/custom-data-ie.service';
// import { FilterAndSortComponent } from '../filter-and-sort/filter-and-sort.component';
import { BsModalService } from '@farris/ui-modal';
import { DataIEUtils } from '../service/Utils/DataIEUtils';
import { HttpClient } from '@angular/common/http';
import { MethodType } from '../model/DataIEType/MethodType';
var ExportDialogComponent = /** @class */ (function () {
    function ExportDialogComponent(dataExportService, notify, fileService, entityService, customIEService, resolver, modalService, http) {
        var _this = this;
        this.dataExportService = dataExportService;
        this.notify = notify;
        this.fileService = fileService;
        this.entityService = entityService;
        this.customIEService = customIEService;
        this.resolver = resolver;
        this.modalService = modalService;
        this.http = http;
        // 规则ID
        this.ruleId = '';
        this.i18nId = '';
        this.exportOption = null;
        this.isHandCraft = false;
        this.subject = null;
        // repository = null;
        this.frameContext = null;
        this.show = false;
        this.needDestroy = new EventEmitter();
        this.valueField = 'Office Excel97-2003(*.xls)';
        this.Model = null;
        this.components = null;
        this.filter = null;
        this.InnerModel = null;
        this.CustomModel = null;
        this.isCustomExport = null;
        // public fileListItems = [
        //     {id:'Office Excel97-2003(*.xls)',label:'Office Excel97-2003(*.xls)'},
        //     {id:'Office Excel(*.xlsx)',label:'Office Excel(*.xlsx)'},
        //     {id:'Office Excel Macro(*.xlsm)',label:'Office Excel Macro(*.xlsm)'},
        //     {id:'WPS Table(*.et)',label:'WPS Table(*.et)'}
        //     // 'WPS Table(Template File)(*.ett)'
        // ];
        this.fileListItems = [
            'Office Excel97-2003(*.xls)',
            'Office Excel(*.xlsx)',
            'Office Excel Macro(*.xlsm)',
            'WPS Table(*.et)'
            // 'WPS Table(Template File)(*.ett)'
        ];
        DataIEUtils.getTranslateValue(http).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            _this.export_success = res['export_success'];
            _this.export_fail = res['export_fail'];
            _this.file_type = res['file_type'];
            _this.download_vsto = res['download_vsto'];
            _this.ie_cancel = res['ie_cancel'];
            _this.ie_ok = res['ie_ok'];
            _this.customexport_no_column = res['customexport_no_column'];
            _this.vsto_component = res['vsto_component'];
        }));
    }
    /**
     * @return {?}
     */
    ExportDialogComponent.prototype.DownloadVSTO = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var repository = this.frameContext === null ? null : this.frameContext.repository;
        if (this.isHandCraft === true) {
            /** @type {?} */
            var su = DataIEUtils.ChooseSu(repository, this.exportOption);
            this.fileService.DownloadFile4HandCraft('数据导出组件.msi', this.vsto_component, su, false);
        }
        else {
            /** @type {?} */
            var befService = DataIEUtils.GetBefRestService(repository);
            this.fileService.DownloadFile('数据导出组件.msi', this.vsto_component, befService.baseUri, false);
        }
    };
    /**
     * @return {?}
     */
    ExportDialogComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        this.needDestroy.emit();
    };
    /**
     * @return {?}
     */
    ExportDialogComponent.prototype.Ok = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.ruleId === '' && this.InnerModel === null) {
            return;
        }
        /** @type {?} */
        var suffix = '.xls';
        switch (this.valueField) {
            case 'Office Excel97-2003(*.xls)':
                suffix = '.xls';
                break;
            case 'Office Excel(*.xlsx)':
                suffix = '.xlsx';
                break;
            case 'Office Excel Macro(*.xlsm)':
                suffix = '.xlsm';
                break;
            case 'WPS Table(*.et)':
                suffix = '.et';
                break;
            case 'WPS Table(Template File)(*.ett)':
                suffix = '.ett';
                break;
            default:
                break;
        }
        if (this.isCustomExport === false) {
            suffix = null;
        }
        //自定义导出分支
        if (this.InnerModel !== null) {
            this.eleSelectComp.Save();
            /** @type {?} */
            var elements = this.InnerModel.MainObject.Elements;
            /** @type {?} */
            var mainObjName = this.InnerModel.MainObject.FileObjectName;
            if (elements == null || elements === undefined || elements.length === 0) {
                this.notify.info(mainObjName + ' ' + this.customexport_no_column);
                return;
            }
            this.customIEService.SaveCustomExportRule(this.InnerModel, this.components, this.frameContext, suffix, this.i18nId, this.exportOption.dimension).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                if (data['success'] === true) {
                    _this.customIEService.ExportByRule(['DataExport', _this.Model['MainObject']['Source']['SourceMetadataID'], _this.exportOption.dimension], _this.filter, _this.exportOption, _this.frameContext);
                }
                else {
                    _this.notify.error(_this.export_fail + data['message']);
                }
            }));
            /** @type {?} */
            var unscriber_1 = this.customIEService.export.subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                if (data === '') {
                    _this.notify.success(_this.export_success);
                    unscriber_1.unsubscribe();
                    _this.subject.next(true);
                    _this.subject.complete();
                }
                else {
                    /** @type {?} */
                    var errorlevel = _this.customIEService.errorlevel;
                    if (errorlevel === 'Info') {
                        _this.notify.info(data);
                    }
                    else if (errorlevel === 'Warning') {
                        _this.notify.warning(data);
                    }
                    else {
                        _this.notify.error(_this.export_fail + data);
                    }
                    // this.notify.error('导出失败！');
                    unscriber_1.unsubscribe();
                    _this.subject.next(false);
                    _this.subject.complete();
                }
                _this.needDestroy.emit();
            }));
        }
        //选VSTO特性下的普通导出分支
        if (this.ruleId !== '') {
            switch (this.methodType) {
                case MethodType.CommonExport:
                    this.dataExportService.ExportById(this.ruleId, this.frameContext, this.exportOption, suffix);
                    /** @type {?} */
                    var unscriber_2 = this.dataExportService.export.subscribe((/**
                     * @param {?} data
                     * @return {?}
                     */
                    function (data) {
                        if (data === '') {
                            _this.notify.success(_this.export_success);
                            unscriber_2.unsubscribe();
                            _this.subject.next(true);
                            _this.subject.complete();
                        }
                        else {
                            /** @type {?} */
                            var errorlevel = _this.dataExportService.errorlevel;
                            if (errorlevel === 'Info') {
                                _this.notify.info(data);
                            }
                            else if (errorlevel === 'Warning') {
                                _this.notify.warning(data);
                            }
                            else {
                                _this.notify.error(_this.export_fail + data);
                            }
                            // this.notify.error('导出失败！');
                            unscriber_2.unsubscribe();
                            _this.subject.next(false);
                            _this.subject.complete();
                        }
                        _this.needDestroy.emit();
                    }));
                    break;
                case MethodType.ExportCurrentTableData:
                    this.dataExportService.ExportCurrentData(this.ruleId, this.frameContext, this.entityService.pageIndex, this.entityService.pageSize, this.exportOption, suffix);
                    /** @type {?} */
                    var exportUnscribe_1 = this.dataExportService.export.subscribe((/**
                     * @param {?} data
                     * @return {?}
                     */
                    function (data) {
                        if (data === '') {
                            _this.notify.success(_this.export_success);
                            exportUnscribe_1.unsubscribe();
                            _this.subject.next(true);
                            _this.subject.complete();
                        }
                        else {
                            /** @type {?} */
                            var errorlevel = _this.dataExportService.errorlevel;
                            if (errorlevel === 'Info') {
                                _this.notify.info(data);
                            }
                            else if (errorlevel === 'Warning') {
                                _this.notify.warning(data);
                            }
                            else {
                                _this.notify.error(_this.export_fail + data);
                            }
                            exportUnscribe_1.unsubscribe();
                            _this.subject.next(false);
                            _this.subject.complete();
                        }
                        _this.needDestroy.emit();
                    }));
                    break;
            }
            // this.needDestroy.emit();
        }
    };
    /**
     * @return {?}
     */
    ExportDialogComponent.prototype.Cancel = /**
     * @return {?}
     */
    function () {
        this.needDestroy.emit(true);
    };
    ExportDialogComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-export-dialog',
                    template: "<div class=\"d-flex flex-column\" style=\"width: 100%; height: 100%;\">\r\n    <div region=\"north\" class=\"north\">\r\n        <div class=\"row\" style=\"margin-top: 12px\">\r\n            <!--\r\n            <div class=\"div-inline\">\r\n                <input type=\"text\" style=\"text-align: center;width: 100px; border: 0px\" name=\"expfiltercond\" id=\"expfiltercond\" value=\"\u8FC7\u6EE4\u6392\u5E8F\u6761\u4EF6\">\r\n            </div>\r\n            <div class=\"div-inline\">\r\n                <input style=\"width:680px\" type=\"text\" id=\"filterExpression\" name=\"filterExpression\" placeholder=\"\u8BBE\u7F6E\u8FC7\u6EE4\u4E0E\u6392\u5E8F\u6761\u4EF6\" />\r\n            </div>\r\n            <div class=\"div-inline\">\r\n                <button class=\" btn btn-primary\" style=\"width:60px\" type=\"button\" (click)=\"SetFilter()\">\u8BBE\u7F6E</button>\r\n            </div>\r\n            -->\r\n            \r\n            <!--<label>\u8FC7\u6EE4</label>\r\n            <filter [columns]=\"columns\", [(ngModel)]=\"filterConditions\"></filter>-->\r\n            <!--\r\n            <label>\u6392\u5E8F</label>\r\n            <sorter [columns]=\"columns\" [(ngModel)]=\"conditions\"></sorter>\r\n            -->\r\n        </div>\r\n        <div style=\"margin-top: 12px\">\r\n            <div class=\"div-inline\">\r\n            </div>\r\n        </div>\r\n    </div>\r\n    <div region=\"center\" class=\"center flex-fill\">\r\n      <div *ngIf=\"isCustomExport\">\r\n          <app-element-select [Model]=\"Model\" [InnerModel]=\"InnerModel\" [CustomModel]=\"CustomModel\"></app-element-select>\r\n        </div>\r\n    </div>\r\n    <div *ngIf=\"isCustomExport\" region=\"south\" class=\"south mgleft\">\r\n        <!-- \u81EA\u5B9A\u4E49\u5BFC\u51FA -->\r\n      <label>{{file_type}}</label>\r\n      <kendo-dropdownlist class=\"longleng\" id=\"fileType\" [data]=\"fileListItems\" [(ngModel)] = \"valueField\">\r\n      </kendo-dropdownlist>\r\n      <!-- <farris-combo-list id=\"fileType\" [(ngModel)]=\"valueField\" [idField]=\"'id'\" [textField]=\"'label'\" [data]=\"fileListItems\"></farris-combo-list> -->\r\n    </div>\r\n</div>\r\n<ng-template #selectFooter>\r\n    <!-- <div class=\"btns\">\r\n        <button type=\"button\" class=\"btn btn-outline-secondary btn-sm px-3\" (click)=\"DownloadVSTO()\" *ngIf=\"show\">{{download_vsto}}</button>\r\n        <button type=\"button\" class=\"btn btn-outline-secondary btn-sm px-3\" (click)=\"Cancel()\">{{ie_cancel}}</button>\r\n        <button type=\"button\" class=\"btn btn-outline-primary btn-sm px-3\" (click)=\"Ok()\">{{ie_ok}}</button>\r\n    </div> -->\r\n    <button type=\"button\" class=\"btn btn-outline-secondary\" (click)=\"DownloadVSTO()\" *ngIf=\"show\">{{download_vsto}}</button>\r\n    <button type=\"button\" class=\"btn btn-outline-secondary\" (click)=\"Cancel()\">{{ie_cancel}}</button>\r\n    <button type=\"button\" class=\"btn btn-primary\" (click)=\"Ok()\">{{ie_ok}}</button>\r\n</ng-template>",
                    providers: [NotifyService],
                    styles: [".barmanager{position:absolute;bottom:0;height:50px;right:30px;padding-left:50px}*{padding:0;margin:0}#filePath,#fileType,#filterExpression{margin-left:7px}.input-group{margin-bottom:12px}.btngroup{margin-top:12px}.exchange{margin-top:30px}#IsSave{margin:6px}.pos{position:relative}.left{float:right}.mkwidth{width:75px;line-height:28px}.div-inline{display:inline}.mgleft{margin-left:20px;margin-top:10px}.longleng{width:820px}"]
                }] }
    ];
    /** @nocollapse */
    ExportDialogComponent.ctorParameters = function () { return [
        { type: DataExportService },
        { type: NotifyService },
        { type: FileOperateService },
        { type: EntityListService },
        { type: CustomDataIEService },
        { type: ComponentFactoryResolver },
        { type: BsModalService },
        { type: HttpClient }
    ]; };
    ExportDialogComponent.propDecorators = {
        modalFooter: [{ type: ViewChild, args: ['selectFooter',] }],
        ruleId: [{ type: Input }],
        i18nId: [{ type: Input }],
        exportOption: [{ type: Input }],
        isHandCraft: [{ type: Input }],
        frameContext: [{ type: Input }],
        methodType: [{ type: Input }],
        show: [{ type: Input }],
        needDestroy: [{ type: Output }],
        eleSelectComp: [{ type: ViewChild, args: [ElementSelectComponent,] }],
        Model: [{ type: Input }],
        components: [{ type: Input }],
        filter: [{ type: Input }],
        InnerModel: [{ type: Input }],
        CustomModel: [{ type: Input }],
        isCustomExport: [{ type: Input }]
    };
    return ExportDialogComponent;
}());
export { ExportDialogComponent };
if (false) {
    /** @type {?} */
    ExportDialogComponent.prototype.modalFooter;
    /** @type {?} */
    ExportDialogComponent.prototype.ruleId;
    /** @type {?} */
    ExportDialogComponent.prototype.i18nId;
    /** @type {?} */
    ExportDialogComponent.prototype.exportOption;
    /** @type {?} */
    ExportDialogComponent.prototype.isHandCraft;
    /** @type {?} */
    ExportDialogComponent.prototype.subject;
    /** @type {?} */
    ExportDialogComponent.prototype.frameContext;
    /** @type {?} */
    ExportDialogComponent.prototype.methodType;
    /** @type {?} */
    ExportDialogComponent.prototype.show;
    /** @type {?} */
    ExportDialogComponent.prototype.needDestroy;
    /** @type {?} */
    ExportDialogComponent.prototype.valueField;
    /** @type {?} */
    ExportDialogComponent.prototype.exportRule;
    /** @type {?} */
    ExportDialogComponent.prototype.eleSelectComp;
    /** @type {?} */
    ExportDialogComponent.prototype.Model;
    /** @type {?} */
    ExportDialogComponent.prototype.components;
    /** @type {?} */
    ExportDialogComponent.prototype.filter;
    /** @type {?} */
    ExportDialogComponent.prototype.InnerModel;
    /** @type {?} */
    ExportDialogComponent.prototype.CustomModel;
    /** @type {?} */
    ExportDialogComponent.prototype.isCustomExport;
    /** @type {?} */
    ExportDialogComponent.prototype.export_success;
    /** @type {?} */
    ExportDialogComponent.prototype.export_fail;
    /** @type {?} */
    ExportDialogComponent.prototype.file_type;
    /** @type {?} */
    ExportDialogComponent.prototype.download_vsto;
    /** @type {?} */
    ExportDialogComponent.prototype.ie_cancel;
    /** @type {?} */
    ExportDialogComponent.prototype.ie_ok;
    /** @type {?} */
    ExportDialogComponent.prototype.customexport_no_column;
    /** @type {?} */
    ExportDialogComponent.prototype.vsto_component;
    /** @type {?} */
    ExportDialogComponent.prototype.fileListItems;
    /**
     * @type {?}
     * @private
     */
    ExportDialogComponent.prototype.dataExportService;
    /**
     * @type {?}
     * @private
     */
    ExportDialogComponent.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    ExportDialogComponent.prototype.fileService;
    /**
     * @type {?}
     * @private
     */
    ExportDialogComponent.prototype.entityService;
    /**
     * @type {?}
     * @private
     */
    ExportDialogComponent.prototype.customIEService;
    /**
     * @type {?}
     * @private
     */
    ExportDialogComponent.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    ExportDialogComponent.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    ExportDialogComponent.prototype.http;
}
