// @farris/mobile-bef/esm2015/lib/bef_variable_manager.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_variable_manager.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetailType } from './types';
import { BefChangeUtil } from './utils/index';
/**
 * Be变量管理器
 */
class BefVariableManager {
    /**
     * 构造函数
     * @param {?} ngVariables
     */
    constructor(ngVariables) {
        this.ngVariables = ngVariables;
        this.ngVariableMap = new Map();
        this.lastSyncValuesMap = new Map();
        this.latestVariableValues = new Map();
        // 重新组织变量元数据
        Object.keys(this.ngVariables).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        (propName) => {
            this.ngVariableMap.set(propName, ngVariables[propName]);
        }));
    }
    /**
     * 设置变量值
     * @param {?} name
     * @param {?} value
     * @return {?}
     */
    setValue(name, value) {
        if (this.ngVariableMap.has(name) === false) {
            throw new Error(`不存在名为${name}的变量`);
        }
        this.latestVariableValues.set(name, value);
    }
    /**
     * 获取变量值
     * @param {?} name
     * @return {?}
     */
    getValue(name) {
        if (this.ngVariableMap.has(name) === false) {
            throw new Error(`不存在名为${name}的变量`);
        }
        this.latestVariableValues.get(name);
    }
    /**
     * 处理服务器端返回的变量变更
     * @param {?} changeDetail
     * @return {?}
     */
    handleChangeDetail(changeDetail) {
    }
    /**
     * Build ChangeDetail instance for all variables.
     * @return {?}
     */
    buildChangeDetail() {
        /** @type {?} */
        const changeDetail = BefChangeUtil.createEmpty(ChangeDetailType.Modify);
        this.ngVariableMap.forEach((/**
         * @param {?} ngVariable
         * @param {?} varName
         * @return {?}
         */
        (ngVariable, varName) => {
            /** @type {?} */
            const newValue = this.latestVariableValues.get(varName);
            /** @type {?} */
            const oldValue = this.lastSyncValuesMap.get(varName);
            if (this.isValueEqual(newValue, oldValue) === false) {
                this.lastSyncValuesMap.set(varName, newValue);
                this.appendToChangeInfo(changeDetail, varName, newValue);
            }
        }));
        if (Object.keys(changeDetail.ChangeInfo).length === 0) {
            return null;
        }
        return changeDetail;
    }
    /**
     * Clear variable values cached in the innerValueMap property.
     * @return {?}
     */
    reset() {
        this.lastSyncValuesMap.clear();
        this.latestVariableValues.clear();
    }
    /**
     * Append changed variable to ChangeDetail instance.
     * @private
     * @param {?} changeDetail
     * @param {?} varName
     * @param {?} varValue
     * @return {?}
     */
    appendToChangeInfo(changeDetail, varName, varValue) {
        if (this.isUdtVariable(varValue) === true) {
            /** @type {?} */
            const udtVarChangeDetail = BefChangeUtil.createEmpty(ChangeDetailType.Modify);
            udtVarChangeDetail.ChangeInfo = varValue;
            changeDetail.ChangeInfo[varName] = udtVarChangeDetail;
        }
        else {
            changeDetail.ChangeInfo[varName] = varValue;
        }
    }
    /**
     * 值比较
     * \@todo 临时采用这种方式
     * @private
     * @param {?} srcValue
     * @param {?} dstValue
     * @return {?}
     */
    isValueEqual(srcValue, dstValue) {
        return JSON.stringify(srcValue) === JSON.stringify(dstValue);
    }
    /**
     * Check if the object is a plain object
     * @private
     * @param {?} obj
     * @return {?}
     */
    isUdtVariable(obj) {
        return obj && obj.constructor &&
            obj.toString() === '[object Object]' &&
            obj.constructor.prototype.hasOwnProperty('isPrototypeOf');
    }
}
if (false) {
    /**
     * 变量元数据
     * @type {?}
     * @private
     */
    BefVariableManager.prototype.ngVariableMap;
    /**
     * 最近向服务器端同步的变量值
     * @type {?}
     * @private
     */
    BefVariableManager.prototype.lastSyncValuesMap;
    /**
     * 当前最新的变量值
     * @type {?}
     * @private
     */
    BefVariableManager.prototype.latestVariableValues;
    /**
     * @type {?}
     * @private
     */
    BefVariableManager.prototype.ngVariables;
}
export { BefVariableManager };
