// @farris/mobile-bef/esm2015/lib/bef_repository_initializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_repository_initializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { MetadataUtil, AppContext, InjectFlags, HttpClient } from '@farris/mobile-devkit';
import { VARIABLE_PROP_META } from './decorators';
import { BefEntityManager } from './bef_entity_manager';
import { BefVariableManager } from './bef_variable_manager';
import { BefSessionHandlingStrategyFactory, BefSessionService } from './session/index';
import { BefDataService } from './bef_data_service';
import { FrameworkSessionService } from './framework_session_service';
import { BE_SESSION_HANDLING_STRATEGY_TOKEN } from './tokens';
import { BefProxyExtend } from './bef_proxy_extend';
/**
 * BefRepository初始化器
 * \@summary
 * 为了抽离这部分路径，强制对外暴露了Repository的相关属性，待优化。
 */
class BefRepositoryInitializer {
    /**
     * 注入器
     * @private
     * @return {?}
     */
    get injector() {
        return this.repository.getInjector();
    }
    /**
     * 构造函数
     * @param {?} repository
     */
    constructor(repository) {
        this.repository = repository;
    }
    /**
     * 初始化
     * @return {?}
     */
    init() {
        this.initAppContext();
        this.initEntityManager();
        this.initVariableManager();
        this.initApiProxy();
        this.initSessionService();
        this.initDataService();
    }
    /**
     * 初始化应用上下文
     * @private
     * @return {?}
     */
    initAppContext() {
        this.repository.appContext = this.injector.get(AppContext);
    }
    /**
     * 初始化EntityManager
     * @private
     * @return {?}
     */
    initEntityManager() {
        this.repository.entityManager = new BefEntityManager(this.repository.entityCollection);
    }
    /**
     * 初始化VariableManager
     * @private
     * @return {?}
     */
    initVariableManager() {
        /** @type {?} */
        const ngVariables = MetadataUtil.getPropsMetadatasByName(this.repository.constructor, VARIABLE_PROP_META);
        this.repository.variableManager = new BefVariableManager(ngVariables);
    }
    /**
     * 初始化Proxy
     * @private
     * @return {?}
     */
    initApiProxy() {
        this.repository.apiProxy = this.injector.get(this.repository.apiProxyType);
        /** @type {?} */
        const apiProxyExtend = new BefProxyExtend(this.repository);
        this.repository.apiProxy.setProxyExtend(apiProxyExtend);
    }
    /**
     * 初始化SessionService
     * @private
     * @return {?}
     */
    initSessionService() {
        /** @type {?} */
        let handlingStrategyName = this.injector.get(BE_SESSION_HANDLING_STRATEGY_TOKEN, null, InjectFlags.Optional);
        handlingStrategyName = handlingStrategyName || 'SeparatedSession';
        /** @type {?} */
        const frmSessionService = this.injector.get(FrameworkSessionService);
        /** @type {?} */
        const httpClient = this.injector.get(HttpClient);
        /** @type {?} */
        const baseUrl = `${this.repository.apiProxy.baseUrl}`;
        /** @type {?} */
        const sessionHandlingStrategyFactory = new BefSessionHandlingStrategyFactory();
        /** @type {?} */
        const sessionHandlingStrategy = sessionHandlingStrategyFactory.create(handlingStrategyName, frmSessionService, baseUrl, httpClient);
        this.repository.sessionService = new BefSessionService(sessionHandlingStrategy);
    }
    /**
     * 初始化DataService
     * @private
     * @return {?}
     */
    initDataService() {
        this.repository.dataService = new BefDataService(this.repository);
    }
}
if (false) {
    /**
     * BefRepository实例
     * @type {?}
     * @private
     */
    BefRepositoryInitializer.prototype.repository;
}
export { BefRepositoryInitializer };
