// @farris/mobile-bef/esm2015/lib/bef_proxy_extend.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_proxy_extend.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { of, EMPTY } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { RequestInfoUtil } from './utils';
import { BE_ERROR_HANDLER__TOKEN } from './tokens';
/**
 * 代理钩子
 * @record
 */
export function IProxyExtend() { }
if (false) {
    /**
     * @param {?} body
     * @return {?}
     */
    IProxyExtend.prototype.extendBody = function (body) { };
    /**
     * @param {?} headers
     * @return {?}
     */
    IProxyExtend.prototype.extendHeaders = function (headers) { };
    /**
     * @param {?} response
     * @param {?=} ignoreChanges
     * @return {?}
     */
    IProxyExtend.prototype.onResponse = function (response, ignoreChanges) { };
    /**
     * @param {?} error
     * @param {?} selfHandError
     * @param {?} ignoreError
     * @return {?}
     */
    IProxyExtend.prototype.onError = function (error, selfHandError, ignoreError) { };
}
/**
 * Bef代理扩展类
 */
export class BefProxyExtend {
    /**
     * 构造函数
     * @param {?} repository
     */
    constructor(repository) {
        this.repository = repository;
    }
    /**
     * 注入器
     * @private
     * @return {?}
     */
    get injector() {
        return this.repository.getInjector();
    }
    /**
     * 返回结果处理
     * @param {?} response
     * @return {?}
     */
    onResponse(response) {
        this.handleResponseHeaders(response.headers);
        return this.handleResponseBody(response.body);
    }
    /**
     * 处理服务器端返回的headers数据
     * @private
     * @param {?} headers
     * @return {?}
     */
    handleResponseHeaders(headers) {
        this.repository.sessionService.handleResponseHeaders(headers);
    }
    /**
     * 处理服务器端返回的的body数据
     * @private
     * @param {?} responseInfo
     * @return {?}
     */
    handleResponseBody(responseInfo) {
        if (responseInfo && responseInfo.innerDataChange) {
            this.repository.entityManager.handleDataChangeDetails(responseInfo.innerDataChange);
        }
        this.repository.entityManager.clearAllEntityChanges();
        if (responseInfo && responseInfo.hasOwnProperty('returnValue')) {
            return responseInfo.returnValue;
        }
        else {
            return responseInfo;
        }
    }
    /**
     * 错误处理
     * @param {?} error
     * @return {?}
     */
    onError(error) {
        // 获取所有load组件实例，消除
        /** @type {?} */
        const loadingServices = window['DEVKIT_LOADING_SERVICE'];
        if (loadingServices && loadingServices instanceof Array && loadingServices.length > 0) {
            for (const loadingService of loadingServices) {
                if (typeof (loadingService.hide) === 'function') {
                    loadingService.hide();
                }
            }
        }
        // 捕获到异常，处理异常信息
        /** @type {?} */
        const exceptionService = this.injector.get(BE_ERROR_HANDLER__TOKEN, null);
        if (!!exceptionService) {
            exceptionService.show(error);
            return EMPTY;
        }
        else {
            return EMPTY;
        }
    }
    /**
     * 扩展Headers
     * @param {?} headers
     * @return {?}
     */
    extendHeaders(headers) {
        /** @type {?} */
        const $getSessionId = this.repository.sessionService.getBeSessionId();
        return $getSessionId.pipe(switchMap((/**
         * @return {?}
         */
        () => {
            headers = this.repository.sessionService.extendRequestHeaders(headers);
            return of(headers);
        })));
    }
    /**
     * 扩展Body
     * @param {?} body
     * @return {?}
     */
    extendBody(body) {
        return RequestInfoUtil.appendRequestInfoToBody(body, this.repository);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    BefProxyExtend.prototype.repository;
}
