// @farris/mobile-bef/esm2015/lib/bef_repository.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_repository.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Repository } from '@farris/mobile-devkit';
import { BefRepositoryInitializer } from './bef_repository_initializer';
/**
 * Bef实体仓库实现类
 * @abstract
 * @template T
 */
class BefRepository extends Repository {
    /**
     * 构造函数
     * @param {?} injector
     */
    constructor(injector) {
        super();
        this.injector = injector;
    }
    /**
     * 初始化
     * \@summary
     * 实现类的构造函数中调用才能正确获取apiProxy
     * @protected
     * @return {?}
     */
    init() {
        super.init();
        /** @type {?} */
        const initializer = new BefRepositoryInitializer(this);
        initializer.init();
    }
    /**
     * 获取注入器
     * @return {?}
     */
    getInjector() {
        return this.injector;
    }
    /**
     * 获取实体数组
     * @param {?} filter
     * @param {?} sort
     * @param {?} pageSize
     * @param {?} pageIndex
     * @return {?}
     */
    getEntities(filter, sort, pageSize, pageIndex) {
        return this.dataService.getList(filter, sort, pageSize, pageIndex);
    }
    /**
     * 获取主键为id的实体
     * @param {?} id 实体id
     * @return {?}
     */
    getEntityById(id) {
        return this.dataService.getById(id);
    }
    /**
     * 更新实体数据
     * @param {?} id
     * @return {?}
     */
    updateEntityById(id) {
        return this.dataService.updateById(id);
    }
    /**
     * 更新实体数据并加锁
     * @param {?} id
     * @return {?}
     */
    editEntityById(id) {
        return this.dataService.editById(id);
    }
    /**
     * 创建实体
     * @param {?=} defaultValue
     * @return {?}
     */
    createEntity(defaultValue) {
        return this.dataService.create(defaultValue);
    }
    /**
     * 追加实体
     * @param {?=} defaultValue
     * @return {?}
     */
    appendEntity(defaultValue) {
        return this.dataService.append(defaultValue);
    }
    /**
     * 创建子实体，并追加
     * @param {?} fpath 父路径，格式形如：/1/edus（从表）或/1/edus/2/grades（从从表）
     * @return {?}
     */
    appendEntityByPath(fpath) {
        return this.dataService.appendByPath(fpath);
    }
    /**
     * 根据id删除实体
     * @param {?} id
     * @param {?=} ifSave
     * @return {?}
     */
    removeEntityById(id, ifSave) {
        return this.dataService.removeById(id, ifSave);
    }
    /**
     * 根据ids批量删除实体
     * @param {?} ids
     * @return {?}
     */
    removeEntitiesByIds(ids) {
        return this.dataService.removeByIds(ids);
    }
    /**
     * 根据id删除实体并执行保存
     * @param {?} id
     * @return {?}
     */
    removeEntityAndSaveById(id) {
        return this.dataService.removeAndSaveById(id);
    }
    /**
     * 删除子级
     * @param {?} fpath
     * @param {?} id
     * @return {?}
     */
    removeEntityByPath(fpath, id) {
        return this.dataService.removeByPath(fpath, id);
    }
    /**
     * 保存主键为id的实体
     * @param {?} id
     * @return {?}
     */
    saveEntityById(id) {
        return this.dataService.applyChangesById(id);
    }
    /**
     * 批量保存
     * @return {?}
     */
    saveEntities() {
        return this.dataService.applyChanges();
    }
    /**
     * 取消实体变更
     * @return {?}
     */
    cancelEntityChanges() {
        return this.dataService.cancelChanges();
    }
    /**
     * 清空Repositoy内的数据和变量
     * @return {?}
     */
    reset() {
        this.entityManager.reset();
        this.variableManager.reset();
        this.sessionService.clearBeSessionId();
    }
}
if (false) {
    /**
     * 名称
     * @type {?}
     */
    BefRepository.prototype.name;
    /**
     * Bef代理
     * @type {?}
     */
    BefRepository.prototype.apiProxyType;
    /**
     * 应用上下文
     * @type {?}
     */
    BefRepository.prototype.appContext;
    /**
     * 实体管理器
     * @type {?}
     */
    BefRepository.prototype.entityManager;
    /**
     * 变量管理器
     * @type {?}
     */
    BefRepository.prototype.variableManager;
    /**
     * API代理
     * @type {?}
     */
    BefRepository.prototype.apiProxy;
    /**
     * Bef会话管理
     * @type {?}
     */
    BefRepository.prototype.sessionService;
    /**
     * Bef数据操作
     * @type {?}
     */
    BefRepository.prototype.dataService;
    /**
     * @type {?}
     * @protected
     */
    BefRepository.prototype.injector;
}
export { BefRepository };
