// @farris/mobile-bef/esm2015/lib/types.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/types.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 变更类型
 */
export class ChangeDetailType {
}
/**
 * 新增
 */
ChangeDetailType.Added = 'Added';
/**
 * 修改
 */
ChangeDetailType.Modify = 'Modify';
/**
 * 删除
 */
ChangeDetailType.Deleted = 'Deleted';
if (false) {
    /**
     * 新增
     * @type {?}
     */
    ChangeDetailType.Added;
    /**
     * 修改
     * @type {?}
     */
    ChangeDetailType.Modify;
    /**
     * 删除
     * @type {?}
     */
    ChangeDetailType.Deleted;
}
/**
 * 行变更信息
 * 必须包含：
 * 1、DataId   => 主键值；
 * 2、属性名   => 新的属性值；
 * 3、子表名+s => 子表行的ChangeDetail数组
 * @record
 */
export function ChangeDetailInfo() { }
if (false) {
    /** @type {?|undefined} */
    ChangeDetailInfo.prototype.DataId;
    /* Skipping unhandled member: [key: string]: number | string | boolean | null | ChangeDetail | ChangeDetail[] | { [key: string]: any };*/
}
/**
 * 行变更详情
 * 包含：
 * 1、变更类型；
 * 2、变更信息
 * @record
 */
export function ChangeDetail() { }
if (false) {
    /** @type {?} */
    ChangeDetail.prototype.ChangeType;
    /** @type {?} */
    ChangeDetail.prototype.ChangeInfo;
}
/**
 * 请求类型
 * @record
 */
export function RequestInfo() { }
if (false) {
    /** @type {?} */
    RequestInfo.prototype.dataChange;
    /** @type {?|undefined} */
    RequestInfo.prototype.variableChange;
}
/**
 * 分页信息
 * @record
 */
export function Pagination() { }
if (false) {
    /** @type {?} */
    Pagination.prototype.pageSize;
    /** @type {?} */
    Pagination.prototype.totalCount;
    /** @type {?} */
    Pagination.prototype.pageCount;
    /** @type {?} */
    Pagination.prototype.pageIndex;
}
/**
 * 查询结果
 * @record
 */
export function QueryResult() { }
if (false) {
    /** @type {?} */
    QueryResult.prototype.result;
    /** @type {?} */
    QueryResult.prototype.pagination;
}
/**
 * 返回结果类型
 * @record
 */
export function ResponseInfo() { }
if (false) {
    /** @type {?} */
    ResponseInfo.prototype.returnValue;
    /** @type {?} */
    ResponseInfo.prototype.message;
    /** @type {?} */
    ResponseInfo.prototype.innerDataChange;
    /** @type {?} */
    ResponseInfo.prototype.innerVariableChange;
}
/**
 * 带RequestInfo的body对象
 * @record
 */
export function BodyWithRequestInfo() { }
if (false) {
    /** @type {?} */
    BodyWithRequestInfo.prototype.requestInfo;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * request option格式
 * @record
 */
export function RequestOption() { }
if (false) {
    /** @type {?} */
    RequestOption.prototype.body;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 *  异常处理信息接口
 * @record
 */
export function IErrorServe() { }
if (false) {
    /**
     * @param {?} error
     * @return {?}
     */
    IErrorServe.prototype.show = function (error) { };
}
