// @gsp-wf/ui-flowchart/esm5/lib/flow-chart/FlowChart.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { mxEvent, mxConstants, mxGraphSelectionModel, mxEventSource } from "@edp-pmf/mxgraph-ts";
import { FlowGraph } from "./FlowGraph";
var FlowChart = /** @class */ (function (_super) {
    tslib_1.__extends(FlowChart, _super);
    function FlowChart(container, config) {
        var _this = _super.call(this) || this;
        _this.mode = 'complete';
        _this.actList = [];
        _this.transInsList = [];
        _this.rootPath = ''; //基路径
        //基路径
        /**
         * 当前语言
         */
        _this.langCode = ((/** @type {?} */ (window))).gspframeworkService && ((/** @type {?} */ (window))).gspframeworkService.rtf.language.getLanguageCode() || localStorage.getItem('languageCode') || 'zh-CHS';
        try {
            _this.rootPath = ((/** @type {?} */ (window))).gspframeworkService.common.getBasePath.get();
        }
        catch (error) {
        }
        _this.container = container;
        _this.createDiv();
        _this.graph = _this.createGraph(config);
        window.setTimeout((/**
         * @return {?}
         */
        function () {
            _this.resetScrollbars();
        }), 0);
        _this.refresh();
        _this.resizeHandler = (/**
         * @return {?}
         */
        function () {
            window.setTimeout((/**
             * @return {?}
             */
            function () {
                if (_this.graph != null) {
                    _this.refresh();
                }
            }), 0);
        });
        mxEvent.addListener(window, 'resize', _this.resizeHandler);
        _this.graph.getSelectionModel().addListener(mxEvent.CHANGE, (/**
         * @param {?} sender
         * @param {?} evt
         * @return {?}
         */
        function (sender, evt) {
            _this.cellSelect(sender, evt);
        }));
        return _this;
    }
    /**
     * 设置当前语言
     * @param lang
     */
    /**
     * 设置当前语言
     * @param {?} lang
     * @return {?}
     */
    FlowChart.prototype.setLanguage = /**
     * 设置当前语言
     * @param {?} lang
     * @return {?}
     */
    function (lang) {
        this.langCode = lang;
        if (this.graph) {
            this.graph.langCode = lang;
        }
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    FlowChart.prototype.switchMode = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var mode = _a.mode, actiInstList = _a.actiInstList, transInsList = _a.transInsList;
        this.mode = mode;
        this.actList = actiInstList;
        this.transInsList = transInsList;
        this.graph.actList = this.actList;
        this.graph.transInsList = this.transInsList;
        this.graph.mode = this.mode;
        this.graph.reloadBpmnModel();
    };
    /**
     * @param {?} data
     * @return {?}
     */
    FlowChart.prototype.load = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data) {
            this.processInstance = data.processInstance;
            this.actList = data.actiInstList;
            this.transInsList = data.transInsList;
            this.mode = data.mode;
            this.graph.actList = this.actList;
            this.graph.transInsList = this.transInsList;
            this.graph.mode = this.mode;
            this.graph.loadBpmnModel(data.content);
            this.resetScrollbars();
        }
    };
    /**
     * @return {?}
     */
    FlowChart.prototype.createDiv = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.diagramContainer = document.createElement('div');
        this.diagramContainer.className = 'flowChartContainer';
        this.container.appendChild(this.diagramContainer);
        setTimeout((/**
         * @return {?}
         */
        function () {
            // this.diagramContainer.style.overflow = 'hidden';
            mxEvent.addListener(_this.diagramContainer, 'mouseover', (/**
             * @return {?}
             */
            function () {
                // this.diagramContainer.style.overflow = 'auto';
                _this.diagramContainer.classList.add('active');
            }));
            mxEvent.addListener(_this.diagramContainer, 'mouseleave', (/**
             * @return {?}
             */
            function () {
                // this.diagramContainer.style.overflow = 'hidden';
                _this.diagramContainer.classList.remove('active');
            }));
        }), 0);
        /** @type {?} */
        var buttonIn = document.createElement('a');
        buttonIn.setAttribute('href', 'javascript:void(0);');
        /** @type {?} */
        var buttonInImg = document.createElement('img');
        buttonIn.style.marginRight = '10px';
        buttonInImg.src = this.rootPath + '/platform/runtime/common/web/@gsp-wf/wf-process-editor/images/zoomIn.svg';
        buttonInImg.style.display = "inline-block";
        buttonInImg.style.verticalAlign = "middle";
        buttonIn.appendChild(buttonInImg);
        mxEvent.addListener(buttonIn, 'click', (/**
         * @param {?} evt
         * @return {?}
         */
        function (evt) {
            _this.graph.zoomIn();
        }));
        /** @type {?} */
        var buttonOut = document.createElement('a');
        buttonOut.setAttribute('href', 'javascript:void(0);');
        /** @type {?} */
        var buttonOutImg = document.createElement('img');
        buttonOutImg.src = this.rootPath + '/platform/runtime/common/web/@gsp-wf/wf-process-editor/images/zoomOut.svg';
        buttonOutImg.style.display = "inline-block";
        buttonOutImg.style.verticalAlign = "middle";
        buttonOut.appendChild(buttonOutImg);
        mxEvent.addListener(buttonOut, 'click', (/**
         * @param {?} evt
         * @return {?}
         */
        function (evt) {
            _this.graph.zoomOut();
        }));
        /** @type {?} */
        var zoomBar = document.createElement('div');
        zoomBar.className = 'zoomBar';
        zoomBar.style.position = 'absolute';
        zoomBar.style.top = '10px';
        zoomBar.style.right = '20px';
        zoomBar.appendChild(buttonIn);
        zoomBar.appendChild(buttonOut);
        // zoomBar.appendChild(zoom);
        this.container.appendChild(zoomBar);
    };
    /**
     * @param {?} config
     * @return {?}
     */
    FlowChart.prototype.createGraph = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        var _this = this;
        mxConstants.VERTEX_SELECTION_STROKEWIDTH = 1;
        mxConstants.VERTEX_SELECTION_COLOR = '#0000CD';
        mxConstants.VERTEX_SELECTION_DASHED = true;
        /** @type {?} */
        var graph = new FlowGraph(this.diagramContainer, config);
        // 创建无限大画布
        graph.createInfiniteCanvas();
        // 创建后需要调用一次才生效
        graph.sizeDidChange();
        graph.popupMenuHandler.autoExpand = true;
        graph.popupMenuHandler.factoryMethod = (/**
         * @param {?} menu
         * @param {?} cell
         * @param {?} evt
         * @return {?}
         */
        function (menu, cell, evt) {
            if (cell != null
                && cell.getId().indexOf('UserActivity') > -1
                && _this.actList.findIndex((/**
                 * @param {?} a
                 * @return {?}
                 */
                function (a) { return a.activityDefinitionId === cell.getId(); })) > -1
                && _this.processInstance) {
                /** @type {?} */
                var actInstId_1 = _this.actList.find((/**
                 * @param {?} a
                 * @return {?}
                 */
                function (a) { return a.activityDefinitionId === cell.getId(); })).activityInstanceId;
                menu.addItem('查看单据', null, (/**
                 * @return {?}
                 */
                function () {
                    _this.viewForm(_this.processInstance.id, actInstId_1, _this.processInstance.bizInstId);
                }));
            }
            if (cell != null
                && cell.getId().indexOf('AifCreationBillActivity') > -1
                && _this.actList.findIndex((/**
                 * @param {?} a
                 * @return {?}
                 */
                function (a) { return a.activityDefinitionId === cell.getId(); })) > -1
                && _this.processInstance) {
                // 筛选状态是Running的自动生单活动实例
                /** @type {?} */
                var actInsts = _this.actList.filter((/**
                 * @param {?} a
                 * @return {?}
                 */
                function (a) { return a.activityDefinitionId === cell.getId() && a.state == "RUNNING"; }));
                if (actInsts.length == 1) {
                    /** @type {?} */
                    var actInst_1 = actInsts[0];
                    menu.addItem('重试', null, (/**
                     * @return {?}
                     */
                    function () {
                        _this.retryAifCreation(_this.processInstance.id, actInst_1.activityInstanceId, _this.processInstance.bizInstId);
                    }));
                }
            }
        });
        return graph;
    };
    /**
     * @param {?} sender
     * @param {?} evt
     * @return {?}
     */
    FlowChart.prototype.cellSelect = /**
     * @param {?} sender
     * @param {?} evt
     * @return {?}
     */
    function (sender, evt) {
        if (!(sender instanceof mxGraphSelectionModel)) {
            return;
        }
        /** @type {?} */
        var selectedCell = sender.cells[0];
        if (selectedCell) {
            if (selectedCell.vertex) {
                /** @type {?} */
                var selectId_1 = selectedCell.getId();
                // 选中子流程节点时，弹出子流程图
                if (selectId_1 && selectId_1.indexOf('CallActivitySubProcess') > -1
                    && this.actList.findIndex((/**
                     * @param {?} a
                     * @return {?}
                     */
                    function (a) { return a.state !== 'FORECAST'
                        && a.activityDefinitionId === selectId_1; })) > -1) {
                    /** @type {?} */
                    var actInsts = this.actList.filter((/**
                     * @param {?} a
                     * @return {?}
                     */
                    function (a) { return a.state !== 'FORECAST' && a.activityDefinitionId === selectId_1; }));
                    /** @type {?} */
                    var activeInst = actInsts.find((/**
                     * @param {?} a
                     * @return {?}
                     */
                    function (a) { return a.state === 'RUNNING'; }));
                    if (activeInst) {
                        this.viewChildProcessChart(activeInst);
                    }
                    else {
                        this.viewChildProcessChart(actInsts[0]);
                    }
                }
                else {
                    this.ngComp.viewLogs({ type: 'activity', actiDefId: selectedCell.id });
                }
            }
            else {
                this.ngComp.viewLogs({ type: 'sequence' });
            }
        }
        else {
            this.ngComp.viewLogs({ type: 'process' });
        }
    };
    /**
     * @param {?} actInst
     * @return {?}
     */
    FlowChart.prototype.viewChildProcessChart = /**
     * @param {?} actInst
     * @return {?}
     */
    function (actInst) {
        this.ngComp.viewChild(actInst, this.processInstance);
    };
    /**
     * @param {?} procInstId
     * @param {?} actInstId
     * @param {?} bizInstId
     * @return {?}
     */
    FlowChart.prototype.viewForm = /**
     * @param {?} procInstId
     * @param {?} actInstId
     * @param {?} bizInstId
     * @return {?}
     */
    function (procInstId, actInstId, bizInstId) {
        this.ngComp.viewForm(procInstId, actInstId, bizInstId);
    };
    /**
     * @param {?} procInstId
     * @param {?} actInstId
     * @param {?} bizInstId
     * @return {?}
     */
    FlowChart.prototype.retryAifCreation = /**
     * @param {?} procInstId
     * @param {?} actInstId
     * @param {?} bizInstId
     * @return {?}
     */
    function (procInstId, actInstId, bizInstId) {
        this.ngComp.retryAifCreation(procInstId, actInstId, bizInstId);
    };
    /**
     * @return {?}
     */
    FlowChart.prototype.resetScrollbars = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var bounds = this.graph.getGraphBounds();
        /** @type {?} */
        var width = Math.max(bounds.width * this.graph.view.scale, 0);
        // this.getScrollTileSize().width * this.graph.view.scale);
        /** @type {?} */
        var height = Math.max(bounds.height * this.graph.view.scale, 0);
        this.graph.container.scrollTop = Math.floor(Math.max(0, bounds.y - Math.max(20, (this.graph.container.clientHeight - height) / 2)));
        this.graph.container.scrollLeft = Math.floor(Math.max(0, bounds.x - Math.max(0, (this.graph.container.clientWidth - width) / 2)));
    };
    /**
     * @return {?}
     */
    FlowChart.prototype.refresh = /**
     * @return {?}
     */
    function () {
        this.resetScrollbars();
    };
    /**
     * @return {?}
     */
    FlowChart.prototype.reloadBpmnModel = /**
     * @return {?}
     */
    function () {
        this.graph.removeCells(this.graph.getChildCells(this.graph.getDefaultParent()));
    };
    return FlowChart;
}(mxEventSource));
export { FlowChart };
if (false) {
    /** @type {?} */
    FlowChart.prototype.ngComp;
    /** @type {?} */
    FlowChart.prototype.container;
    /** @type {?} */
    FlowChart.prototype.graph;
    /** @type {?} */
    FlowChart.prototype.diagramContainer;
    /** @type {?} */
    FlowChart.prototype.button;
    /** @type {?} */
    FlowChart.prototype.resizeHandler;
    /** @type {?} */
    FlowChart.prototype.mode;
    /** @type {?} */
    FlowChart.prototype.processInstance;
    /** @type {?} */
    FlowChart.prototype.bpmnModel;
    /** @type {?} */
    FlowChart.prototype.actList;
    /** @type {?} */
    FlowChart.prototype.transInsList;
    /** @type {?} */
    FlowChart.prototype.rootPath;
    /**
     * 当前语言
     * @type {?}
     */
    FlowChart.prototype.langCode;
}
