// @edp-pmf/bpmn-model/dist/esm/lib/model/participant/PalParticipant.js
import { __extends } from "tslib";
import { ExtensionMethod } from "../ExtensionMethod";
import { OrganizationLevel } from "./OrganizationLevel";
import { PalParticipantType } from "./PalParticipantType";
import { Participant } from "./Participant";
var PalParticipant = /** @class */ (function (_super) {
    __extends(PalParticipant, _super);
    function PalParticipant(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "pal.PositionParticipant";
        _this.palParticipantType = PalParticipantType.Position;
        // 组织
        _this.organizationLevel = OrganizationLevel.DEPARTMENT;
        _this.positionOrgType = "Sys";
        return _this;
    }
    PalParticipant.prototype.NewInstance = function () {
        return new PalParticipant(this.Model);
    };
    PalParticipant.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.palParticipantType = PalParticipantType.Position;
        this.organizationLevel = OrganizationLevel.DEPARTMENT;
        this.organizationId = null;
        this.organizationCode = null;
        this.organizationName = null;
        this.positionId = null;
        this.positionCode = null;
        this.positionName = null;
        this.positionType = null;
        this.positionOrgId = null;
        this.positionOrgName = null;
        this.positionOrgType = null;
        this.userId = null;
        this.userCode = null;
        this.userName = null;
    };
    PalParticipant.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "PalParticipantType", this.palParticipantType);
        ExtensionMethod.SetAttributeValue(obj, "OrganizationLevel", this.organizationLevel);
        ExtensionMethod.SetAttributeValue(obj, "OrganizationId", this.organizationId);
        ExtensionMethod.SetAttributeValue(obj, "OrganizationCode", this.organizationCode);
        ExtensionMethod.SetAttributeValue(obj, "OrganizationName", this.organizationName);
        ExtensionMethod.SetAttributeValue(obj, "PositionId", this.positionId);
        ExtensionMethod.SetAttributeValue(obj, "PositionCode", this.positionCode);
        ExtensionMethod.SetAttributeValue(obj, "PositionName", this.positionName);
        ExtensionMethod.SetAttributeValue(obj, "PositionType", this.positionType);
        ExtensionMethod.SetAttributeValue(obj, "PositionOrgId", this.positionOrgId);
        ExtensionMethod.SetAttributeValue(obj, "PositionOrgName", this.positionOrgName);
        ExtensionMethod.SetAttributeValue(obj, "PositionOrgType", this.positionOrgType);
        ExtensionMethod.SetAttributeValue(obj, "UserId", this.userId);
        ExtensionMethod.SetAttributeValue(obj, "UserCode", this.userCode);
        ExtensionMethod.SetAttributeValue(obj, "UserName", this.userName);
        return obj;
    };
    PalParticipant.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.palParticipantType = PalParticipantType[ExtensionMethod.GetAttributeValue4String(obj, "PalParticipantType")];
        this.organizationLevel = OrganizationLevel[ExtensionMethod.GetAttributeValue4String(obj, "OrganizationLevel")];
        this.organizationId = ExtensionMethod.GetAttributeValue4String(obj, "OrganizationId");
        this.organizationCode = ExtensionMethod.GetAttributeValue4String(obj, "OrganizationCode");
        this.organizationName = ExtensionMethod.GetAttributeValue4String(obj, "OrganizationName");
        this.positionId = ExtensionMethod.GetAttributeValue4String(obj, "PositionId");
        this.positionCode = ExtensionMethod.GetAttributeValue4String(obj, "PositionCode");
        this.positionName = ExtensionMethod.GetAttributeValue4String(obj, "PositionName");
        this.positionType = ExtensionMethod.GetAttributeValue4String(obj, "PositionType");
        this.positionOrgId = ExtensionMethod.GetAttributeValue4String(obj, "PositionOrgId");
        this.positionOrgName = ExtensionMethod.GetAttributeValue4String(obj, "PositionOrgName");
        this.positionOrgType = ExtensionMethod.GetAttributeValue4String(obj, "PositionOrgType");
        this.userId = ExtensionMethod.GetAttributeValue4String(obj, "UserId");
        this.userCode = ExtensionMethod.GetAttributeValue4String(obj, "UserCode");
        this.userName = ExtensionMethod.GetAttributeValue4String(obj, "UserName");
    };
    PalParticipant.prototype.clone = function () {
        var c = this.NewInstance();
        c.palParticipantType = this.palParticipantType;
        c.organizationId = this.organizationId;
        c.organizationCode = this.organizationCode;
        c.organizationName = this.organizationName;
        c.organizationLevel = this.organizationLevel;
        c.positionId = this.positionId;
        c.positionCode = this.positionCode;
        c.positionName = this.positionName;
        c.positionOrgId = this.positionOrgId;
        c.positionOrgName = this.positionOrgName;
        c.positionOrgType = this.positionOrgType;
        c.positionType = this.positionType;
        c.userId = this.userId;
        c.userCode = this.userCode;
        c.userName = this.userName;
        return c;
    };
    return PalParticipant;
}(Participant));
export { PalParticipant };
