// @farris/mobile-devkit/esm5/lib/ui-state/control-state/control_state_interceptor.service.js
import * as tslib_1 from "tslib";
import { shareReplay } from "rxjs/operators";
import { InjectFlags } from "../../core";
import { CONTROL_STATE_INTERCEPTOR_TOKEN, InitContext, ProcessContext } from "./control_state_interceptor";
import { forkJoin, isObservable, of } from "rxjs";
import { EnvUtil, UrlUtil } from "../../utils";
/**
 * 状态拦截器服务
 */
var ControlStateInterceptorService = /** @class */ (function () {
    function ControlStateInterceptorService(injector) {
        this.injector = injector;
        //控件状态拦截器
        this.controlInterceptors = [];
        this.controlInterceptors = this.injector.get(CONTROL_STATE_INTERCEPTOR_TOKEN, null, InjectFlags.Optional);
    }
    ControlStateInterceptorService.prototype.init = function () {
        var e_1, _a;
        if (this.result) {
            return this.result;
        }
        var inits = [];
        var initContext = new InitContext();
        var processContext = this.getProcessContext();
        initContext.processContext = this.processContext = processContext;
        try {
            for (var _b = tslib_1.__values(this.controlInterceptors), _c = _b.next(); !_c.done; _c = _b.next()) {
                var controlInterceptor = _c.value;
                if (controlInterceptor.init && typeof controlInterceptor.init === 'function') {
                    var initFunc = controlInterceptor.init(initContext);
                    var $result = isObservable(initFunc) ? initFunc : of(initFunc);
                    inits.push($result);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return this.result = forkJoin(inits).pipe(shareReplay(1));
    };
    ControlStateInterceptorService.prototype.intercept = function (interceptContext) {
        var e_2, _a;
        interceptContext.processContext = this.processContext;
        try {
            for (var _b = tslib_1.__values(this.controlInterceptors), _c = _b.next(); !_c.done; _c = _b.next()) {
                var controlInterceptor = _c.value;
                controlInterceptor.intercept(interceptContext);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    ControlStateInterceptorService.prototype.getProcessContext = function () {
        var processContext = new ProcessContext();
        var isInWf = EnvUtil.isInWf();
        processContext.isInWf = isInWf;
        if (isInWf === false) {
            return processContext;
        }
        var iframe = window.parent.document.querySelector('iframe');
        if (!iframe || !iframe.src) {
            return processContext;
        }
        var queryParams = UrlUtil.getParams(iframe.src);
        if (!queryParams) {
            return processContext;
        }
        processContext.uiStateInProcess = queryParams.UIStateInProcess;
        processContext.formConfigId = queryParams.formConfigId;
        return processContext;
    };
    return ControlStateInterceptorService;
}());
export { ControlStateInterceptorService };
