// @farris/mobile-devkit/esm5/lib/ui-state/control-state/control_state_listener.js
import { ChangeType } from "../../binding-data";
import { ControlStateUpdater } from "./control_state_updater";
var ControlStateListener = /** @class */ (function () {
    function ControlStateListener(viewModelContext) {
        var _this = this;
        //监听订阅集合
        this.subscriptionsMap = new Map();
        this.viewModelContext = viewModelContext;
        this.controlStateUpdater = viewModelContext.injector.get(ControlStateUpdater);
        //更新器初始化完成后，注册监听,主动触发一次更新
        this.controlStateUpdater.init().subscribe(function (res) {
            _this.register();
            var changes = {
                lintenerType: 'firstUpdate',
                change: null
            };
            _this.controlStateUpdater.updateControlState(changes);
        });
    }
    /**
     *注册监听
     */
    ControlStateListener.prototype.register = function () {
        this.registerBindingDataLintener();
        this.registerUIStateLintener();
        this.registerStateMachineLintener();
        this.registerFormLintener();
    };
    /**
     * 监听BindingData变化
     */
    ControlStateListener.prototype.registerBindingDataLintener = function () {
        var _this = this;
        var stateName = 'bindingData';
        if (this.subscriptionsMap.has(stateName) === false) {
            var subscription = this.viewModelContext.bindingData.changes.subscribe(function (change) {
                if (change.type === ChangeType.Load) {
                    return;
                }
                _this.controlStateUpdater.updateControlState({
                    lintenerType: 'bindingData',
                    change: change
                });
            });
            this.subscriptionsMap.set(stateName, subscription);
        }
    };
    /**
     * 监听UIState变化
     */
    ControlStateListener.prototype.registerUIStateLintener = function () {
        var _this = this;
        var stateName = 'uiState';
        if (this.subscriptionsMap.has(stateName) === false) {
            var subscription = this.viewModelContext.uiState.changes.subscribe(function (change) {
                _this.controlStateUpdater.updateControlState({
                    lintenerType: 'uiState',
                    change: change
                });
            });
            this.subscriptionsMap.set(stateName, subscription);
        }
    };
    /**
     * 监听StateMachine变化
     */
    ControlStateListener.prototype.registerStateMachineLintener = function () {
        var _this = this;
        var stateName = 'stateMachine';
        if (this.subscriptionsMap.has(stateName) === false && this.viewModelContext.stateMachine) {
            var subscription = this.viewModelContext.stateMachine.stateChange.subscribe(function (change) {
                _this.controlStateUpdater.updateControlState({
                    lintenerType: 'stateMachine',
                    change: change
                });
            });
            this.subscriptionsMap.set(stateName, subscription);
        }
    };
    /**
     * 监听Form变化
     */
    ControlStateListener.prototype.registerFormLintener = function () {
        var _this = this;
        var stateName = 'form';
        if (this.subscriptionsMap.has(stateName) === false) {
            var subscription = this.viewModelContext.form.changes.subscribe(function (change) {
                if (change && change.type === 'validateFieldsFinished') {
                    _this.controlStateUpdater.updateControlState({
                        lintenerType: 'expression',
                        change: change
                    });
                }
                else {
                    _this.controlStateUpdater.updateControlState({
                        lintenerType: 'form',
                        change: change
                    });
                }
            });
            this.subscriptionsMap.set(stateName, subscription);
        }
    };
    return ControlStateListener;
}());
export { ControlStateListener };
