// @farris/devkit/esm5/lib/rest/rest_service.js
/* eslint-disable new-cap */
/* eslint-disable valid-jsdoc */
/* eslint-disable require-jsdoc */
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { BasePathService } from '@farris/rtf';
var RestfulService = /** @class */ (function () {
    /**
     * 构造函数
     * @param httpClient http客户端
     */
    function RestfulService(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * 发送GET请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    RestfulService.prototype.get = function (uri, params, options) {
        return this.request(uri, 'GET', params, options);
    };
    /**
     * 发送POST请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    RestfulService.prototype.put = function (uri, body, params, options) {
        var optionsWithBody = this.addBody(options, body);
        return this.request(uri, 'PUT', params, optionsWithBody);
    };
    /**
     * 发送PUT请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    RestfulService.prototype.post = function (uri, body, params, options) {
        var optionsWithBody = this.addBody(options, body);
        return this.request(uri, 'POST', params, optionsWithBody);
    };
    /**
     * 发送DELETE请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    RestfulService.prototype.delete = function (uri, params, options) {
        return this.request(uri, 'DELETE', params, options);
    };
    /**
     * 发送请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    RestfulService.prototype.request = function (uri, method, params, options) {
        if (options === void 0) { options = {}; }
        options = options || {};
        if (params) {
            var httpParams = this.buildParams(params);
            options.params = httpParams;
        }
        var methodName = method;
        uri = BasePathService.convertPath(uri);
        var result = this.httpClient.request(methodName, uri, options);
        return result;
    };
    /**
     * 构造Http请求参数
     * @param params 请求参数
     */
    RestfulService.prototype.buildParams = function (params) {
        var httpParams = new HttpParams();
        for (var key in params) {
            if (params.hasOwnProperty(key)) {
                var value = params[key].toString();
                httpParams = httpParams.append(key, value);
            }
        }
        return httpParams;
    };
    RestfulService.prototype.addBody = function (options, body) {
        options = options || {};
        var mergedOptions = Object.assign(options, { body: body });
        return mergedOptions;
    };
    RestfulService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    RestfulService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return RestfulService;
}());
export { RestfulService };
