// @farris/ui-list-filter/esm2015/lib/entity/conditionvalue/checkboxgroupvalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
export class CheckboxGroupValue {
    /**
     * @param {?=} value
     */
    constructor(value = []) {
        this.clear$ = new EventEmitter();
        this.valueChanged = new EventEmitter();
        this.value = value;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return this.value.length == 0;
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.value = [];
        this.clear$.emit();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        this.value = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setValue(value) {
        this.value = value;
        this.valueChanged.emit(this.value);
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.value;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.CheckboxGroup;
    }
}
if (false) {
    /** @type {?} */
    CheckboxGroupValue.prototype.clear$;
    /** @type {?} */
    CheckboxGroupValue.prototype.valueChanged;
    /** @type {?} */
    CheckboxGroupValue.prototype.value;
}
