// @farris/ui-list-filter/esm2015/lib/entity/conditionvalue/dropdownlistvalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
import { EnumValueConvert } from "../controltype/dropdownlist/ienumvalue";
import { EventEmitter } from "@angular/core";
export class DropDownListValue {
    /**
     * @param {?=} data
     */
    constructor(data = '') {
        this.clear$ = new EventEmitter();
        this.value = data;
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.value = undefined;
        this.clear$.emit();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        this.value = value;
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.value;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.DropDown;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return !this.value;
    }
}
if (false) {
    /** @type {?} */
    DropDownListValue.prototype.clear$;
    /** @type {?} */
    DropDownListValue.prototype.value;
}
export class DropDownListValueConvert {
    /**
     * @param {?} obj
     * @return {?}
     */
    convertToObject(obj) {
        /** @type {?} */
        let dropDownListValue = (/** @type {?} */ (obj));
        /** @type {?} */
        let jobj = new Object();
        /** @type {?} */
        let convertor = new EnumValueConvert();
        if (!!(dropDownListValue.value)) {
            // jobj["value"]=convertor.convertJObject(dropDownListValue.value);
            jobj["value"] = dropDownListValue.value;
        }
        return jobj;
    }
    /**
     * @param {?} jobj
     * @return {?}
     */
    initFromObject(jobj) {
        /** @type {?} */
        let dropDownListValue = new DropDownListValue();
        if (!!(jobj["value"])) {
            /** @type {?} */
            let convertor = new EnumValueConvert();
            // dropDownListValue.value=convertor.initFromJobject(jobj["value"]);
            dropDownListValue.value = jobj["value"];
        }
        return dropDownListValue;
    }
}
