// @farris/ui-list-filter/esm2015/lib/entity/conditionvalue/month.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
import { EventEmitter } from "@angular/core";
export class MonthValue {
    /**
     * @param {?=} date
     */
    constructor(date = '') {
        this.clear$ = new EventEmitter();
        this.value = date;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        this.value = value.formatted;
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.value;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.Date;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return !this.value;
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.value = undefined;
        this.clear$.emit();
    }
}
if (false) {
    /** @type {?} */
    MonthValue.prototype.value;
    /** @type {?} */
    MonthValue.prototype.clear$;
}
