// @farris/ui-list-filter/esm2015/lib/entity/conditionvalue/radiovalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
import { EventEmitter } from "@angular/core";
export class RadioGroupValue {
    /**
     * @param {?=} data
     */
    constructor(data = '') {
        this.clear$ = new EventEmitter();
        this.value = data;
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.value = undefined;
        this.clear$.emit();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        throw new Error("Method not implemented.");
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.value;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.Radio;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return !this.value;
    }
    /**
     * @return {?}
     */
    getValueText() {
        return this.value;
    }
}
if (false) {
    /** @type {?} */
    RadioGroupValue.prototype.value;
    /** @type {?} */
    RadioGroupValue.prototype.clear$;
}
