// @farris/ui-list-filter/esm2015/lib/entity/conditionvalue/textvalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
export class TextValue {
    /**
     * @param {?=} data
     */
    constructor(data = '') {
        this.clear$ = new EventEmitter();
        this.value = data;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.Text;
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.value = undefined;
        this.clear$.emit();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        throw new Error('Method not implemented.');
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.value;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return !this.value;
    }
}
if (false) {
    /** @type {?} */
    TextValue.prototype.value;
    /** @type {?} */
    TextValue.prototype.clear$;
}
