// @farris/kendo-binding/esm5/lib/directives/farris-datagrid/custom-size.directive.js
import { ChangeDetectorRef, Directive, ElementRef, Injector, Input } from '@angular/core';
import { DatagridComponent } from '@farris/ui-datagrid';
import { FrameContext } from '@farris/devkit';
var CustomDatagridSizeDirective = /** @class */ (function () {
    function CustomDatagridSizeDirective(dg, injector, cd, el, frameContext) {
        this.dg = dg;
        this.injector = injector;
        this.cd = cd;
        this.el = el;
        this.frameContext = frameContext;
        this.useCustomSize = false;
        this.sizeApiUri = '';
    }
    CustomDatagridSizeDirective.prototype.restService = function () {
        return this.frameContext.repository.restService;
    };
    CustomDatagridSizeDirective.prototype.ngOnInit = function () {
    };
    CustomDatagridSizeDirective.prototype.ngOnChanges = function (changes) {
        if (changes.sizeApiUri && !changes.sizeApiUri.isFirstChange()) {
            this.applyNewSize();
        }
    };
    CustomDatagridSizeDirective.prototype.ngAfterViewInit = function () {
        this.applyNewSize();
    };
    CustomDatagridSizeDirective.prototype.applyNewSize = function () {
        var _this = this;
        if (this.useCustomSize && this.sizeApiUri) {
            this.getSizeConfig().subscribe(function (cfg) {
                if (_this.dg && typeof _this.dg.applyNewSize === 'function') {
                    _this.dg.applyNewSize(cfg);
                }
            });
        }
    };
    CustomDatagridSizeDirective.prototype.getSizeConfig = function () {
        return this.restService().request(this.sizeApiUri, 'get');
    };
    CustomDatagridSizeDirective.decorators = [
        { type: Directive, args: [{ selector: '[custom-size]' },] }
    ];
    /** @nocollapse */
    CustomDatagridSizeDirective.ctorParameters = function () { return [
        { type: DatagridComponent },
        { type: Injector },
        { type: ChangeDetectorRef },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    CustomDatagridSizeDirective.propDecorators = {
        useCustomSize: [{ type: Input, args: ['custom-size',] }],
        sizeApiUri: [{ type: Input }]
    };
    return CustomDatagridSizeDirective;
}());
export { CustomDatagridSizeDirective };
