// @farris/kendo-binding/esm5/lib/directives/farris-datagrid/remote-summary.directive.js
import { Directive, Input } from '@angular/core';
import { ViewModel } from '@farris/devkit';
import { DatagridComponent } from '@farris/ui-datagrid';
var FarrisDataGridRemoteSummaryDirective = /** @class */ (function () {
    function FarrisDataGridRemoteSummaryDirective(viewModel, grid) {
        var _this = this;
        this.viewModel = viewModel;
        this.grid = grid;
        var appContext = this.viewModel.frameContext.getFormAppContext();
        if (appContext && appContext.hasOwnProperty('messagePipe')) {
            appContext.messagePipe.subscribe(function (message) {
                var messageType = message && message.type || null;
                // 查询事件
                if (messageType === 'query') {
                    if (_this.remoteSummaryCommand && _this.grid.footerDataFrom === 'server') {
                        var paths = _this.remoteSummaryCommand.split('.');
                        var func_1 = null;
                        paths.forEach(function (path) {
                            func_1 = func_1 && func_1[path] || _this[path];
                        });
                        if (typeof func_1 === 'function') {
                            func_1().subscribe(function (result) {
                                _this.grid.footerData = result;
                            });
                        }
                    }
                }
            });
        }
    }
    /**
     * 执行命令
     * @param commandName 命令名称
     * @param frameId frameId
     */
    FarrisDataGridRemoteSummaryDirective.prototype.execute = function (commandName, frameId) {
        if (!commandName || commandName === '' || commandName === 'undefined') {
            return;
        }
        var viewModel = this.viewModel;
        if (frameId) {
            var targetContext = viewModel.frameContext.appContext.frameContextManager.getFrameContextById(frameId);
            viewModel = targetContext.viewModel;
        }
        return viewModel[commandName]();
    };
    FarrisDataGridRemoteSummaryDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[farris-remote-summary-command]"
                },] }
    ];
    /** @nocollapse */
    FarrisDataGridRemoteSummaryDirective.ctorParameters = function () { return [
        { type: ViewModel },
        { type: DatagridComponent }
    ]; };
    FarrisDataGridRemoteSummaryDirective.propDecorators = {
        remoteSummaryCommand: [{ type: Input, args: ['farris-remote-summary-command',] }]
    };
    return FarrisDataGridRemoteSummaryDirective;
}());
export { FarrisDataGridRemoteSummaryDirective };
