// @farris/mobile-devkit/esm5/lib/expression/expression/expression_result_factory.js
import { ExpressionManager } from "./expression_manager";
;
import { ExpressionRegistry } from "./expression_registry";
import { ExpressionResult } from "./expression_result";
var ExpressionResultFactory = /** @class */ (function () {
    function ExpressionResultFactory(viewModelContext) {
        this.viewModelContext = viewModelContext;
        this.injector = this.viewModelContext.injector;
        this.expressionRegistry = this.injector.get(ExpressionRegistry);
        this.expressionManager = this.injector.get(ExpressionManager);
        this.expressionResult = this.injector.get(ExpressionResult);
        this.registeEvent();
    }
    ExpressionResultFactory.prototype.registeEvent = function () {
        var _this = this;
        this.expressionRegistry.expressions.subscribe(function (expressionObjects) {
            // 加载完表达式之后做一次计算
            expressionObjects.forEach(function (expressionObject) {
                if (expressionObject.deps && expressionObject.deps.length > 0) {
                    return;
                }
                var result = _this.expressionManager.eval(expressionObject.id);
                _this.expressionResult[expressionObject.id] = result;
            });
        });
    };
    return ExpressionResultFactory;
}());
export { ExpressionResultFactory };
