// @farris/mobile-devkit/esm5/lib/expression/expression/expression_registry.js
import * as tslib_1 from "tslib";
import { of } from "rxjs";
import { catchError, switchMap } from "rxjs/operators";
import { TranslateToken } from "../../i18n/index";
import { FORM_EXPRESSION_MANIFEST_SERVICE_TOKEN } from "../manifest/index";
import { Expression } from './types';
import { AppContext } from '../../app';
var ExpressionRegistry = /** @class */ (function () {
    function ExpressionRegistry(viewModelContext) {
        this.viewModelContext = viewModelContext;
        this._expressions = null;
        this.injector = this.viewModelContext.injector;
        this.formExpressionManifestService = this.injector.get(FORM_EXPRESSION_MANIFEST_SERVICE_TOKEN, null);
        this.translate = this.injector.get(TranslateToken, null);
    }
    /**
     * 加载表达式文件
     */
    ExpressionRegistry.prototype.load = function () {
        var _this = this;
        return this.formExpressionManifestService.load().pipe(switchMap(function (describe) {
            var expressions = _this.setExpressions(describe);
            _this.cleanSpecialCharacters();
            return of(expressions);
        }), catchError(function (e) {
            return of([]);
        }));
    };
    ExpressionRegistry.prototype.setExpressions = function (describe) {
        var _this = this;
        var expressions = [];
        var exprs = Array.from(describe);
        var appContext = this.viewModelContext.injector.get(AppContext);
        exprs.forEach(function (expr) {
            expr.expressions.forEach(function (expression) {
                var _a, _b;
                var expressionObject = {
                    id: expression.id,
                    ns: expr.ns,
                    viewModelId: expr.viewModelId,
                    path: expr.path,
                    bindingType: expr.type,
                    type: expression.type,
                    expression: expression.value || expression.expr || '',
                    message: expression.message || null,
                    messageType: expression.messageType || null,
                    deps: []
                };
                if (expression.type === Expression.ExpressionType.Required) {
                    var b = { type: "requiredExpression", constraints: [expression.id], message: expression.message };
                    appContext.validationManager[expr.path] = tslib_1.__assign({}, (appContext.validationManager[expr.path] || {}), (_a = {}, _a[expression.id] = b, _a));
                }
                if (expression.type === Expression.ExpressionType.Validate) {
                    var b = { type: "expression", constraints: [expression.id], message: expression.message };
                    appContext.validationManager[expr.path] = tslib_1.__assign({}, (appContext.validationManager[expr.path] || {}), (_b = {}, _b[expression.id] = b, _b));
                }
                if ((expression.type === Expression.ExpressionType.Required || expression.type === Expression.ExpressionType.Validate)) {
                    if (!expression.message) {
                        expressionObject.message = _this.getExpressionMessage(expression.type);
                    }
                    if (!expression.messageType) {
                        expressionObject.messageType = 'error';
                    }
                }
                if (expressionObject.message) {
                    _this.transform(expressionObject);
                }
                expressions.push(expressionObject);
            });
        });
        this._expressions = expressions;
        return expressions;
    };
    Object.defineProperty(ExpressionRegistry.prototype, "expressions", {
        /**
         * 获取所有表达式
         */
        get: function () {
            if (this._expressions) {
                return of(this._expressions);
            }
            // 短路，实现功能
            if (window['__expressions__']) {
                this.setExpressions(window['__expressions__']);
                return of(this._expressions);
            }
            return this.load();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 根据表达式id获取对应的表达式对象
     * @param id 表达式id
     * @returns
     */
    ExpressionRegistry.prototype.getExpressionById = function (id) {
        if (!this._expressions || this._expressions.length < 1) {
            return null;
        }
        return this._expressions.find(function (expressionObject) { return expressionObject.id === id; });
    };
    ExpressionRegistry.prototype.getExpressionMessage = function (expressionType, defaultValue) {
        if (!(expressionType === Expression.ExpressionType.Validate || expressionType === Expression.ExpressionType.Required)) {
            return null;
        }
        if (!this.translate) {
            return defaultValue;
        }
        var currentLanguage = this.translate.getCurrentLanguage() || 'zh-CHS';
        return Expression.MESSAGE[currentLanguage][expressionType];
    };
    ExpressionRegistry.prototype.transform = function (expressionObject) {
        if (!this.translate) {
            return;
        }
        if (expressionObject.message && expressionObject.message.startsWith('{{') && expressionObject.message.endsWith('}}')) {
            expressionObject.message = this.translate.transform(expressionObject.message.substr(2, expressionObject.message.length - 4), null) || this.getExpressionMessage(expressionObject.type);
        }
    };
    ExpressionRegistry.prototype.cleanSpecialCharacters = function () {
        var _this = this;
        if (!this._expressions || this._expressions.length < 1 || !Array.isArray(this._expressions)) {
            return;
        }
        var repository = this.viewModelContext.repository;
        if (!repository) {
            return;
        }
        var entityTypeInfo = repository.entityTypeInfo;
        var regex = new RegExp("[\\'\\\"]?\\s*(" + entityTypeInfo.entityInfo.nodeCode + "|" + entityTypeInfo.entityInfo.originalCode + ")[\\.\\[\\]a-zA-Z0-9_]+\\s*[\\'\\\"]?", 'g');
        this._expressions.forEach(function (expressionObject) {
            var expr = expressionObject.expression;
            var entityPropertyExpressions = expr.match(regex);
            if (Array.isArray(entityPropertyExpressions) && entityPropertyExpressions.length > 0) {
                // 解析出所有实体相关的字符串，以主实体名字开头，包含主实体属性或子表
                entityPropertyExpressions.forEach(function (item) {
                    if (item.indexOf('.') === -1) {
                        console.warn("\u65E0\u6548\u7684\u5B9E\u4F53\u8868\u8FBE\u5F0F:" + item);
                        return;
                    }
                    // 去数组
                    if (/\[\d\]/g.test(item)) {
                        var replacer = item.replace(/\[\d\]/g, '');
                        expressionObject.expression = _this.replaceAll(expressionObject.expression, item, replacer);
                    }
                    // 去星号
                    if (/\*/g.test(item)) {
                        var replacer = item.replace(/\*/g, '');
                        expressionObject.expression = _this.replaceAll(expressionObject.expression, item, replacer);
                    }
                });
            }
        });
    };
    ExpressionRegistry.prototype.replaceAll = function (originalValue, search, replacer) {
        return originalValue.split(search).join(replacer);
    };
    return ExpressionRegistry;
}());
export { ExpressionRegistry };
