// @farris/mobile-devkit/esm5/lib/expression/expression/expression_event_emitter.js
import * as tslib_1 from "tslib";
import { BehaviorSubject } from "rxjs";
import { Listeners } from "../listener/index";
/**
 * 事件监听器主要有以下几种类型的事件：
 * 1、值变化
 * 2、行切换
 *    行切换使用场景为需要为数据计算依赖当前行时
 * 3、数据加载
 */
var ExpressionEventEmitter = /** @class */ (function () {
    function ExpressionEventEmitter(viewModelContext) {
        var _this = this;
        this.viewModelContext = viewModelContext;
        this.listeners = this.viewModelContext.injector.get(Listeners);
        this.events = new Array();
        this.listeners.onEvent.subscribe(function (eventArgs) {
            if (_this.onEvent && _this.onEvent.observers.length > 0) {
                var events = [];
                if (_this.events.length > 0) {
                    events = tslib_1.__spread(_this.events);
                }
                events.push(eventArgs);
                _this.onEvent.next(events);
                _this.events = [];
            }
            else {
                _this.events.push(eventArgs);
            }
        });
    }
    ExpressionEventEmitter.prototype.attach = function () {
        if (!this.onEvent) {
            this.onEvent = new BehaviorSubject(this.events);
        }
        return this.onEvent.asObservable();
    };
    return ExpressionEventEmitter;
}());
export { ExpressionEventEmitter };
