// @gsp-wf/wf-process-model/esm2015/lib/model/participants/ActivityExecutorParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { ParticipantExpression } from "./ParticipantExpression";
export class ActivityExecutorParticipantExpression extends ParticipantExpression {
    /**
     * @param {?} model
     * @param {?=} activityDefId
     * @param {?=} activityDefName
     */
    constructor(model, activityDefId, activityDefName) {
        super(model);
        this.ClrTypeID = ActivityExecutorParticipantExpression.Clr_Type_ID;
        this.activityDefId = activityDefId;
        this.activityDefName = activityDefName;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "ActivityDefId", this.activityDefId);
        BpmnModelHelper.SetAttributeValue(obj, "ActivityDefName", this.activityDefName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.activityDefId = ExtensionMethod.GetAttributeValue4String(obj, "ActivityDefId");
        this.activityDefName = ExtensionMethod.GetAttributeValue4String(obj, "ActivityDefName");
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ActivityExecutorParticipantExpression(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const activityExecutorParticipant = new ActivityExecutorParticipantExpression(this.Model);
        activityExecutorParticipant.id = BpmnModelHelper.GenerateElementId();
        activityExecutorParticipant.name = this.name;
        activityExecutorParticipant.activityDefId = this.activityDefId;
        activityExecutorParticipant.activityDefName = this.activityDefName;
        // if (this.Condition) {
        //     activityExecutorParticipant.Condition = this.Condition.clone();
        // }
        return activityExecutorParticipant;
    }
}
ActivityExecutorParticipantExpression.Clr_Type_ID = 'wf.ActivityExecutorParticipantExpression';
if (false) {
    /** @type {?} */
    ActivityExecutorParticipantExpression.Clr_Type_ID;
    /** @type {?} */
    ActivityExecutorParticipantExpression.prototype.ClrTypeID;
    /** @type {?} */
    ActivityExecutorParticipantExpression.prototype.activityDefId;
    /** @type {?} */
    ActivityExecutorParticipantExpression.prototype.activityDefName;
}
BpmnModelHelper.RegisterElementType(new ActivityExecutorParticipantExpression(new BpmnModel()));
