// @gsp-wf/wf-process-model/esm2015/lib/model/participants/ComponentParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { SerializeUtils } from "../utils/SerializeUtils";
import { ParticipantExpression } from "./ParticipantExpression";
export class ComponentParticipantExpression extends ParticipantExpression {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = ComponentParticipantExpression.Clr_Type_ID;
    }
    /**
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.actualParameters = new Array();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "ComponentId", this.componentId);
        BpmnModelHelper.SetAttributeValue(obj, "ComponentName", this.componentName);
        BpmnModelHelper.SetAttributeValue(obj, "MethodCode", this.methodCode);
        BpmnModelHelper.SetAttributeValue(obj, "MethodName", this.methodName);
        SerializeUtils.setListAttributeValue(obj, 'ActualParameters', this.actualParameters);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.componentId = obj.ComponentId;
        this.componentName = obj.ComponentName;
        this.methodCode = obj.MethodCode;
        this.methodName = obj.MethodName;
        this.actualParameters = SerializeUtils.getListFromJsonObject(this.Model, obj, 'ActualParameters');
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ComponentParticipantExpression(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const componentParticipant = new ComponentParticipantExpression(this.Model);
        componentParticipant.id = BpmnModelHelper.GenerateElementId();
        componentParticipant.name = this.name;
        componentParticipant.componentId = this.componentId;
        componentParticipant.componentName = this.componentName;
        componentParticipant.methodCode = this.methodCode;
        componentParticipant.methodName = this.methodName;
        componentParticipant.actualParameters = new Array();
        if (this.actualParameters && this.actualParameters.length > 0) {
            for (const p of this.actualParameters) {
                /** @type {?} */
                const parameter = p.clone();
                componentParticipant.actualParameters.push(parameter);
            }
        }
        // if (this.Condition) {
        //     componentParticipant.Condition = this.Condition.clone();
        // }
        return componentParticipant;
    }
}
ComponentParticipantExpression.Clr_Type_ID = 'wf.ComponentParticipantExpression';
ComponentParticipantExpression.CLR_TYPE_NAME = '构件';
if (false) {
    /** @type {?} */
    ComponentParticipantExpression.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    ComponentParticipantExpression.CLR_TYPE_NAME;
    /** @type {?} */
    ComponentParticipantExpression.prototype.ClrTypeID;
    /** @type {?} */
    ComponentParticipantExpression.prototype.componentId;
    /** @type {?} */
    ComponentParticipantExpression.prototype.componentName;
    /** @type {?} */
    ComponentParticipantExpression.prototype.methodCode;
    /** @type {?} */
    ComponentParticipantExpression.prototype.methodName;
    /** @type {?} */
    ComponentParticipantExpression.prototype.actualParameters;
}
BpmnModelHelper.RegisterElementType(new ComponentParticipantExpression(new BpmnModel()));
