// @gsp-wf/wf-process-model/esm2015/lib/model/participants/ParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { ActionScope } from "../ActionScope";
import { Condition } from "../extension";
/**
 * @abstract
 */
export class ParticipantExpression extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.actionScope = ActionScope.Default;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        ExtensionMethod.SetAttributeValue(obj, "ActionScope", this.actionScope ? this.actionScope : ActionScope.Default);
        ExtensionMethod.SetAttributeValue(obj, "Id", this.id);
        ExtensionMethod.SetAttributeValue(obj, "Name", this.name);
        ExtensionMethod.SetAttributeValue(obj, "Order", this.order);
        if (this.condition) {
            ExtensionMethod.SetAttributeValue(obj, "Condition", this.condition.ConvertToJsonObject());
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.actionScope = (/** @type {?} */ ((ExtensionMethod.GetAttributeValue4String(obj, "ActionScope") ? ExtensionMethod.GetAttributeValue4String(obj, "ActionScope") : ActionScope.Default)));
        this.id = ExtensionMethod.GetAttributeValue4String(obj, "Id");
        this.name = ExtensionMethod.GetAttributeValue4String(obj, "Name");
        this.order = ExtensionMethod.GetAttributeValue4Int64(obj, "Order");
        if (obj.Condition) {
            /** @type {?} */
            let condition = new Condition(this.Model);
            condition.LoadFromJsonObject(obj.Condition);
            this.condition = condition;
        }
    }
}
if (false) {
    /** @type {?} */
    ParticipantExpression.prototype.id;
    /** @type {?} */
    ParticipantExpression.prototype.name;
    /** @type {?} */
    ParticipantExpression.prototype.actionScope;
    /** @type {?} */
    ParticipantExpression.prototype.condition;
    /** @type {?} */
    ParticipantExpression.prototype.order;
}
