// @gsp-wf/wf-process-model/esm2015/lib/model/participants/ParticipantSetting.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { AutomaticApprovalRule } from "../AutomaticApprovalRule";
export class ParticipantSetting extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = ParticipantSetting.Clr_Type_ID;
        this.MexTag = 'ParticipantSetting';
        this.NoPerformerStrategy = 'leave';
        this.HistoricFirst = false;
    }
    /**
     * @protected
     * @return {?}
     */
    resetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.NoPerformerStrategy = 'leave';
        this.HistoricFirst = true;
        this.ExclusionPerformers = new Array();
        this.automaticPerformers = new Array();
        this.automaticApprovalRules = new Array();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "ExclusionPerformers", this.ExclusionPerformers);
        BpmnModelHelper.SetAttributeValue(obj, "AutomaticPerformers", this.automaticPerformers);
        BpmnModelHelper.SetAttributeValue(obj, "NoPerformerStrategy", this.NoPerformerStrategy);
        BpmnModelHelper.SetAttributeValue(obj, "HistoricFirst", this.HistoricFirst);
        BpmnModelHelper.SetAttributeValue(obj, "Neighboring", this.Neighboring);
        if (this.automaticApprovalRules && this.automaticApprovalRules.length > 0) {
            /** @type {?} */
            const rules = new Array();
            for (const a of this.automaticApprovalRules) {
                rules.push(a.ConvertToJsonObject());
            }
            BpmnModelHelper.SetAttributeValue(obj, "AutomaticApprovalRules", rules);
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.ExclusionPerformers = obj.ExclusionPerformers;
        this.NoPerformerStrategy = obj.NoPerformerStrategy;
        this.HistoricFirst = obj.HistoricFirst;
        this.Neighboring = obj.Neighboring;
        this.automaticPerformers = ExtensionMethod.GetAttributeValue(obj, "AutomaticPerformers");
        this.automaticApprovalRules = new Array();
        /** @type {?} */
        const rules = ExtensionMethod.GetAttributeValue(obj, "AutomaticApprovalRules");
        if (rules && rules.length > 0) {
            for (const r of rules) {
                /** @type {?} */
                const automaticApprovalRule = new AutomaticApprovalRule(this.Model);
                automaticApprovalRule.LoadFromJsonObject(r);
                this.automaticApprovalRules.push(automaticApprovalRule);
            }
        }
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ParticipantSetting(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const participantSetting = new ParticipantSetting(this.Model);
        participantSetting.ExclusionPerformers = new Array();
        if (this.ExclusionPerformers && this.ExclusionPerformers.length > 0) {
            for (const e of this.ExclusionPerformers) {
                participantSetting.ExclusionPerformers.push(e);
            }
        }
        participantSetting.NoPerformerStrategy = this.NoPerformerStrategy;
        participantSetting.HistoricFirst = this.HistoricFirst;
        participantSetting.Neighboring = this.Neighboring;
        participantSetting.automaticPerformers = new Array();
        if (this.automaticPerformers && this.automaticPerformers.length > 0) {
            for (const a of this.automaticPerformers) {
                participantSetting.automaticPerformers.push(a);
            }
        }
        participantSetting.automaticApprovalRules = new Array();
        if (this.automaticApprovalRules && this.automaticApprovalRules.length > 0) {
            for (const r of this.automaticApprovalRules) {
                /** @type {?} */
                const automaticApprovalRule = r.clone();
                participantSetting.automaticApprovalRules.push(automaticApprovalRule);
            }
        }
        return participantSetting;
    }
}
ParticipantSetting.Clr_Type_ID = "wf.ParticipantSetting";
if (false) {
    /** @type {?} */
    ParticipantSetting.Clr_Type_ID;
    /** @type {?} */
    ParticipantSetting.prototype.ClrTypeID;
    /** @type {?} */
    ParticipantSetting.prototype.MexTag;
    /**
     * 参与者排除项
     * 场景示例：流程发起人不参与审批、已审人员不参与审批、上一活动审批人不参与审批、利益相关人（取自表单字段：报销人、排除本单位。。。假象的）
     * 为了保持扩展性，此处引用参与者Id，使用参与者解析器解析
     * @type {?}
     */
    ParticipantSetting.prototype.ExclusionPerformers;
    /**
     * 找不到参与者时处理策略：throwException, leave
     * @type {?}
     */
    ParticipantSetting.prototype.NoPerformerStrategy;
    /**
     * 再次进入该节点时，参与人为该节点的已办人员
     * 主要用于驳回和重新提交至驳回节点的场景
     * @type {?}
     */
    ParticipantSetting.prototype.HistoricFirst;
    /**
     * 连岗审批：同一办理人在相邻节点任务仅办理一次，下一环节自动完成
     * 与exclusionPerformers设置“上一活动审批人不参与审批”冲突
     * @type {?}
     */
    ParticipantSetting.prototype.Neighboring;
    /**
     * 自动审批项
     * 场景示例：流程发起人不参与审批、已审人员不参与审批、上一活动审批人不参与审批、利益相关人（取自表单字段：报销人、排除本单位。。。假象的）
     * 为了保持扩展性，此处引用参与者Id，使用参与者解析器解析
     * @type {?}
     */
    ParticipantSetting.prototype.automaticPerformers;
    /**
     * 自动审批规则List
     * @type {?}
     */
    ParticipantSetting.prototype.automaticApprovalRules;
}
BpmnModelHelper.RegisterElementType(new ParticipantSetting(new BpmnModel()));
