// @gsp-wf/wf-process-model/esm2015/lib/model/participants/ParticipantScheme.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement } from "@edp-pmf/bpmn-model";
import { Condition } from "../extension/condition-expression/Condition";
export class ParticipantScheme extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = ParticipantScheme.Clr_Type_ID;
        this.MexTag = 'participantScheme';
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ParticipantScheme(this.Model);
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "Id", this.Id);
        BpmnModelHelper.SetAttributeValue(obj, "Name", this.Name);
        BpmnModelHelper.SetAttributeValue(obj, "Performers", this.Performers);
        if (this.Condition) {
            BpmnModelHelper.SetAttributeValue(obj, "Condition", this.Condition.ConvertToJsonObject());
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.Id = obj.Id;
        this.Name = obj.Name;
        this.Performers = obj.Performers;
        /** @type {?} */
        const conditionElement = obj.Condition;
        if (conditionElement) {
            /** @type {?} */
            const condition = new Condition(this.Model);
            condition.LoadFromJsonObject(conditionElement);
            this.Condition = condition;
        }
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.Id = '';
        this.Name = '';
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const participantScheme = new ParticipantScheme(this.Model);
        participantScheme.Id = this.Id;
        participantScheme.Name = this.Name;
        participantScheme.Performers = this.Performers;
        if (this.Condition) {
            participantScheme.Condition = this.Condition.clone();
        }
        return participantScheme;
    }
}
ParticipantScheme.Clr_Type_ID = "wf.ParticipantScheme";
if (false) {
    /** @type {?} */
    ParticipantScheme.Clr_Type_ID;
    /** @type {?} */
    ParticipantScheme.prototype.ClrTypeID;
    /** @type {?} */
    ParticipantScheme.prototype.MexTag;
    /** @type {?} */
    ParticipantScheme.prototype.Id;
    /** @type {?} */
    ParticipantScheme.prototype.Name;
    /** @type {?} */
    ParticipantScheme.prototype.Performers;
    /** @type {?} */
    ParticipantScheme.prototype.Condition;
}
BpmnModelHelper.RegisterElementType(new ParticipantScheme(new BpmnModel()));
