// @gsp-wf/wf-process-model/esm2015/lib/model/participants/PositionParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { ParticipantExpression } from "./ParticipantExpression";
export class PositionParticipantExpression extends ParticipantExpression {
    /**
     * @param {?} model
     * @param {?=} positionId
     * @param {?=} positionName
     */
    constructor(model, positionId, positionName) {
        super(model);
        this.ClrTypeID = PositionParticipantExpression.Clr_Type_ID;
        /// <summary>
        /// 组织类型Code
        /// </summary>
        this.OrgType = 'Sys';
        this.PositionId = positionId;
        this.PositionName = positionName;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "PositionId", this.PositionId);
        BpmnModelHelper.SetAttributeValue(obj, "PositionName", this.PositionName);
        BpmnModelHelper.SetAttributeValue(obj, "PositionType", this.PositionType);
        BpmnModelHelper.SetAttributeValue(obj, "OrgId", this.OrgId);
        BpmnModelHelper.SetAttributeValue(obj, "OrgName", this.OrgName);
        BpmnModelHelper.SetAttributeValue(obj, "OrgType", this.OrgType);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.PositionId = obj.PositionId;
        this.PositionName = obj.PositionName;
        this.PositionType = obj.PositionType;
        this.OrgId = obj.OrgId;
        this.OrgName = obj.OrgName;
        this.OrgType = obj.OrgType;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new PositionParticipantExpression(this.Model);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const positionParticipant = new PositionParticipantExpression(this.Model);
        positionParticipant.id = BpmnModelHelper.GenerateElementId();
        positionParticipant.name = this.name;
        positionParticipant.PositionId = this.PositionId;
        positionParticipant.PositionName = this.PositionName;
        positionParticipant.PositionType = this.PositionType;
        positionParticipant.OrgType = this.OrgType;
        positionParticipant.OrgId = this.OrgId;
        positionParticipant.OrgName = this.OrgName;
        // if (this.Condition) {
        //     positionParticipant.Condition = this.Condition.clone();
        // }
        return positionParticipant;
    }
}
PositionParticipantExpression.Clr_Type_ID = 'wf.PositionParticipantExpression';
PositionParticipantExpression.CLR_TYPE_NAME = '岗位';
if (false) {
    /** @type {?} */
    PositionParticipantExpression.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    PositionParticipantExpression.CLR_TYPE_NAME;
    /** @type {?} */
    PositionParticipantExpression.prototype.ClrTypeID;
    /** @type {?} */
    PositionParticipantExpression.prototype.PositionId;
    /** @type {?} */
    PositionParticipantExpression.prototype.PositionName;
    /** @type {?} */
    PositionParticipantExpression.prototype.PositionType;
    /** @type {?} */
    PositionParticipantExpression.prototype.OrgId;
    /** @type {?} */
    PositionParticipantExpression.prototype.OrgName;
    /** @type {?} */
    PositionParticipantExpression.prototype.OrgType;
}
BpmnModelHelper.RegisterElementType(new PositionParticipantExpression(new BpmnModel()));
