// @gsp-wf/wf-process-model/esm2015/lib/model/participants/UserParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { ParticipantExpression } from "./ParticipantExpression";
export class UserParticipantExpression extends ParticipantExpression {
    /**
     * @param {?} model
     * @param {?=} userId
     * @param {?=} name
     * @param {?=} orgType
     */
    constructor(model, userId, name, orgType) {
        super(model);
        this.ClrTypeID = UserParticipantExpression.Clr_Type_ID;
        this.OrgType = 'Sys';
        this.UserId = userId;
        this.name = name;
        this.OrgType = orgType ? orgType : 'Sys';
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new UserParticipantExpression(this.Model);
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "UserId", this.UserId);
        BpmnModelHelper.SetAttributeValue(obj, "OrgId", this.OrgId);
        BpmnModelHelper.SetAttributeValue(obj, "OrgName", this.OrgName);
        BpmnModelHelper.SetAttributeValue(obj, "OrgType", this.OrgType);
        BpmnModelHelper.SetAttributeValue(obj, "Tag", this.Tag);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.OrgId = obj.OrgId;
        this.OrgName = obj.OrgName;
        this.OrgType = obj.OrgType;
        this.UserId = obj.UserId;
        this.Tag = obj.Tag;
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const userParticipant = new UserParticipantExpression(this.Model);
        userParticipant.id = BpmnModelHelper.GenerateElementId();
        userParticipant.name = this.name;
        userParticipant.OrgId = this.OrgId;
        userParticipant.OrgName = this.OrgName;
        userParticipant.OrgType = this.OrgType;
        userParticipant.UserId = this.UserId;
        userParticipant.Tag = this.Tag;
        // if (this.Condition) { 
        //     userParticipant.Condition = this.Condition.clone();
        // }
        return userParticipant;
    }
}
UserParticipantExpression.Clr_Type_ID = "wf.UserParticipantExpression";
UserParticipantExpression.CLR_TYPE_NAME = '用户';
if (false) {
    /** @type {?} */
    UserParticipantExpression.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    UserParticipantExpression.CLR_TYPE_NAME;
    /** @type {?} */
    UserParticipantExpression.prototype.ClrTypeID;
    /** @type {?} */
    UserParticipantExpression.prototype.Tag;
    /** @type {?} */
    UserParticipantExpression.prototype.UserId;
    /** @type {?} */
    UserParticipantExpression.prototype.OrgId;
    /** @type {?} */
    UserParticipantExpression.prototype.OrgName;
    /** @type {?} */
    UserParticipantExpression.prototype.OrgType;
}
BpmnModelHelper.RegisterElementType(new UserParticipantExpression(new BpmnModel()));
