// @gsp-wf/wf-process-model/esm2015/lib/model/participants/VariableParticipantExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { ParticipantExpression } from "./ParticipantExpression";
export class VariableParticipantExpression extends ParticipantExpression {
    /**
     * @param {?} model
     * @param {?=} varValue
     * @param {?=} name
     * @param {?=} varType
     */
    constructor(model, varValue, name, varType) {
        super(model);
        this.ClrTypeID = VariableParticipantExpression.Clr_Type_ID;
        this.VarValue = varValue;
        this.name = name;
        this.VarType = varType ? varType : 'Sys';
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new VariableParticipantExpression(this.Model);
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "Name", this.Name);
        BpmnModelHelper.SetAttributeValue(obj, "Tag", this.Tag);
        BpmnModelHelper.SetAttributeValue(obj, "VarValue", this.VarValue);
        BpmnModelHelper.SetAttributeValue(obj, "VarType", this.VarType);
        BpmnModelHelper.SetAttributeValue(obj, "Relation", this.Relation);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.Name = obj.Name;
        this.Tag = obj.Tag;
        this.VarValue = obj.VarValue;
        this.VarType = obj.VarType;
        this.Relation = obj.Relation;
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const variableParticipant = new VariableParticipantExpression(this.Model);
        variableParticipant.id = BpmnModelHelper.GenerateElementId();
        variableParticipant.Name = this.Name;
        variableParticipant.Tag = this.Tag;
        variableParticipant.VarType = this.VarType;
        variableParticipant.VarValue = this.VarValue;
        variableParticipant.Relation = this.Relation;
        return variableParticipant;
    }
}
VariableParticipantExpression.Clr_Type_ID = "wf.VariableParticipantExpression";
VariableParticipantExpression.CLR_TYPE_NAME = '变量参与者';
if (false) {
    /** @type {?} */
    VariableParticipantExpression.Clr_Type_ID;
    /**
     * @type {?}
     * @private
     */
    VariableParticipantExpression.CLR_TYPE_NAME;
    /** @type {?} */
    VariableParticipantExpression.prototype.ClrTypeID;
    /** @type {?} */
    VariableParticipantExpression.prototype.VarValue;
    /** @type {?} */
    VariableParticipantExpression.prototype.VarType;
    /** @type {?} */
    VariableParticipantExpression.prototype.Relation;
    /** @type {?} */
    VariableParticipantExpression.prototype.Tag;
    /** @type {?} */
    VariableParticipantExpression.prototype.Name;
}
BpmnModelHelper.RegisterElementType(new VariableParticipantExpression(new BpmnModel()));
