// @gsp-svc/file-viewer/esm5/lib/file-list/file-list.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ImageViewerModule } from '../image-viewer/image-viewer.module';
import { DocViewerModule } from '../doc-viewer/doc-viewer.module';
import { OfdViewerModule } from '../ofd-viewer/ofd-viewer.module';
import { HttpService, SessionService } from '@ecp-caf/caf-common';
import { LoadingService } from '@farris/ui-loading';
import { LayoutModule } from '@progress/kendo-angular-layout';
import { ButtonsModule } from '@progress/kendo-angular-buttons';
import { FileListComponent } from './file-list/file-list.component';
import { PdfViewerModule } from '../pdf-viewer/pdf-viewer.module';
import { RotationDirective } from '../service/rotation.directive';
import { SearchService } from '../service/search.service';
import { SelectionService } from '../service/selection.service';
import { ZoomService } from '../service/zoom.service';
import { ConfigService } from '../service/config.service';
import { DocViewerService } from '../service/doc-viewer.service';
import { FileListService } from '../service/file-list.service';
import { NavigateService } from '../service/navigate.service';
import { PagePreloadService } from '../service/page-preload.service';
import { WindowService as ws } from '../service/window.service';
import { CloudprintService, ɵa, ɵb, ɵc } from '@gsp-svc/cloudprint';
import { AppService } from '@gsp-sys/rtf-common';
import { WindowContainerService, WindowService } from '@progress/kendo-angular-dialog';
import { PptViewerModule } from '../ppt-viewer/ppt-viewer.module';
var FileListModule = /** @class */ (function () {
    function FileListModule() {
    }
    FileListModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [FileListComponent],
                    imports: [
                        CommonModule,
                        ImageViewerModule,
                        DocViewerModule,
                        OfdViewerModule,
                        LayoutModule,
                        ButtonsModule,
                        PdfViewerModule,
                        PptViewerModule
                    ],
                    exports: [FileListComponent],
                    entryComponents: [FileListComponent],
                    providers: [HttpService, SessionService, LoadingService, RotationDirective,
                        SearchService, SelectionService, ZoomService, ConfigService, DocViewerService,
                        FileListService, NavigateService, PagePreloadService, ws, CloudprintService,
                        ɵa, ɵb, ɵc, AppService, WindowService, WindowContainerService]
                },] }
    ];
    return FileListModule;
}());
export { FileListModule };
