// @farris/ui-datepicker/esm5/lib/components/dynamic-view/dynamic-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, ChangeDetectorRef, Component, Injector, Input, Output } from '@angular/core';
var DynamicViewComponent = /** @class */ (function () {
    function DynamicViewComponent(injector, cd) {
        this.injector = injector;
        this.cd = cd;
        this._value = '';
        this._data = null;
        this.previousMaxDays = 30;
        this.futureMaxDays = 30;
        this.previous_dateType = 'day';
        this.previous_days = 1;
        this.future_dateType = 'day';
        this.future_days = 1;
        this.present_first_dateType = 'month';
        this.present_last_dateType = 'month';
        this.valueChange = new EventEmitter();
    }
    Object.defineProperty(DynamicViewComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._value = val;
            this.valueChange.emit(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DynamicViewComponent.prototype, "data", {
        get: /**
         * @return {?}
         */
        function () {
            return this._data;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._data = val;
            this.setValue(val);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DynamicViewComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} $event
     * @return {?}
     */
    DynamicViewComponent.prototype.onListItemFormControlClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var lbl = $event.target.closest('label');
        if (lbl) {
            lbl.click();
        }
    };
    /**
     * @param {?} val
     * @return {?}
     */
    DynamicViewComponent.prototype.setValue = /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        if (val) {
            this.value = val.name;
            switch (this.value) {
                case 'present_day': // 今天
                case 'present_month': // 本月
                case 'present_year':
                    break;
                case 'previou':
                    this.previous_dateType = val.dateType;
                    this.previous_days = val.count;
                    break;
                case 'future':
                    this.future_dateType = val.dateType;
                    this.future_days = val.count;
                    break;
                case 'present_first':
                    this.present_first_dateType = val.dateType;
                    break;
                case 'present_last':
                    this.present_last_dateType = val.dateType;
                    break;
            }
            this.cd.detectChanges();
        }
    };
    /**
     * @return {?}
     */
    DynamicViewComponent.prototype.getValue = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var val = null;
        if (this.value) {
            switch (this.value) {
                case 'present_day': // 今天
                    val = { dateType: 'day', period: 'present', count: null, isFirstDay: false, isLastDay: false };
                    break;
                case 'present_month': // 本月
                    val = { dateType: 'month', period: 'present', count: null, isFirstDay: false, isLastDay: false };
                    break;
                case 'present_year':
                    val = { dateType: 'year', period: 'present', count: null, isFirstDay: false, isLastDay: false };
                    break;
                case 'previou':
                    val = { dateType: this.previous_dateType, period: 'previous', count: this.previous_days, isFirstDay: false, isLastDay: false };
                    break;
                case 'future':
                    val = { dateType: this.future_dateType, period: 'future', count: this.future_days, isFirstDay: false, isLastDay: false };
                    break;
                case 'present_first':
                    val = { dateType: this.present_first_dateType, period: 'present', count: null, isFirstDay: true, isLastDay: false };
                    break;
                case 'present_last':
                    val = { dateType: this.present_last_dateType, period: 'present', count: null, isFirstDay: false, isLastDay: true };
                    break;
            }
            if (val) {
                val.name = this.value;
            }
        }
        return val;
    };
    /**
     * @private
     * @param {?} dt
     * @return {?}
     */
    DynamicViewComponent.prototype.getDateTypeString = /**
     * @private
     * @param {?} dt
     * @return {?}
     */
    function (dt) {
        switch (dt) {
            case 'week':
                return this.locale.week || '天';
            case 'day':
                return this.locale.day || '天';
            case 'month':
                return this.locale.month || '月';
            case 'year':
                return this.locale.year || '年';
        }
    };
    /**
     * @return {?}
     */
    DynamicViewComponent.prototype.formatValue = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var val = this.getValue();
        if (val) {
            /** @type {?} */
            var fval = '';
            switch (val.name) {
                case 'present_day': // 今天
                    fval = this.locale.today || '今天';
                    break;
                case 'present_month': // 本月
                    fval = this.locale.presentMonth || '本月';
                    break;
                case 'present_year':
                    fval = this.locale.presentYear || '本年';
                    break;
                case 'previou':
                    fval = (this.locale.previous || '过去') + ' ' + val.count + ' ' + this.getDateTypeString(val.dateType);
                    break;
                case 'future':
                    fval = (this.locale.future || '未来') + ' ' + val.count + ' ' + this.getDateTypeString(val.dateType);
                    break;
                case 'present_first':
                    fval = (this.locale.current || '本') + this.getDateTypeString(val.dateType) + (this.locale.first || '第 1 天');
                    break;
                case 'present_last':
                    fval = (this.locale.current || '本') + this.getDateTypeString(val.dateType) + (this.locale.last || '最后 1 天');
                    break;
            }
            return fval;
        }
        else {
            return '';
        }
    };
    /**
     * @private
     * @param {?} typ
     * @param {?} days
     * @return {?}
     */
    DynamicViewComponent.prototype.setDays = /**
     * @private
     * @param {?} typ
     * @param {?} days
     * @return {?}
     */
    function (typ, days) {
        if (typ === 'previous') {
            this.previous_days = days;
        }
        else {
            this.future_days = days;
        }
    };
    /**
     * @param {?} $event
     * @param {?} tye
     * @return {?}
     */
    DynamicViewComponent.prototype.checkMaxDays = /**
     * @param {?} $event
     * @param {?} tye
     * @return {?}
     */
    function ($event, tye) {
        /** @type {?} */
        var target = (/** @type {?} */ ($event.target));
        /** @type {?} */
        var val = target.value.replace(/[^\d]/g, '');
        /** @type {?} */
        var days = Number(val);
        target.value = days;
        /** @type {?} */
        var max = tye === 'previous' ? this.previousMaxDays : this.futureMaxDays;
        this.setDays(tye, days);
        if (days > max) {
            target.value = max;
            this.setDays(tye, max);
            return;
        }
        else {
            if (days < 1) {
                target.value = 1;
                this.setDays(tye, 1);
                return;
            }
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DynamicViewComponent.prototype.onNumberKeydown = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        $event.stopPropagation();
        if ($event.key === '.' || $event.key === '-' || $event.key === '+') {
            $event.preventDefault();
        }
    };
    DynamicViewComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-dynamic-view',
                    template: "<div class=\"datepicker_dynamic-view\">\r\n    <ul>\r\n        <li class=\"dynamic-view-item\" [class.active]=\"value === 'present_day'\">  \r\n            <div class=\"custom-control custom-radio\">\r\n                <input class=\"custom-control-input\" type=\"radio\" name=\"dynamic_date\" id=\"radio_dynamic_today\" value=\"present_day\" tabindex=\"0\" [(ngModel)]=\"value\">\r\n                <label class=\"custom-control-label\" for=\"radio_dynamic_today\">{{locale?.today}}</label>\r\n            </div>\r\n        </li>\r\n        <!--\r\n        <li class=\"dynamic-view-item\" [class.active]=\"value === 'present_week'\">\r\n            <div class=\"custom-control custom-radio\">\r\n                <input class=\"custom-control-input\" type=\"radio\" name=\"dynamic_date\" id=\"radio_dynamic_present_week\" value=\"present_week\" tabindex=\"0\" [(ngModel)]=\"value\">\r\n                <label class=\"custom-control-label\" for=\"radio_dynamic_present_week\">\u672C\u5468</label>\r\n            </div>\r\n        </li>\r\n        -->\r\n        <li class=\"dynamic-view-item\" [class.active]=\"value === 'present_month'\">\r\n            <div class=\"custom-control custom-radio\">\r\n            <input class=\"custom-control-input\" type=\"radio\" name=\"dynamic_date\" id=\"radio_dynamic_present_month\" value=\"present_month\" tabindex=\"0\" [(ngModel)]=\"value\">\r\n            <label class=\"custom-control-label\" for=\"radio_dynamic_present_month\">{{locale?.presentMonth}}</label>\r\n        </div></li>\r\n        <li class=\"dynamic-view-item\" [class.active]=\"value === 'present_year'\">\r\n            <div class=\"custom-control custom-radio\">\r\n            <input class=\"custom-control-input\" type=\"radio\" name=\"dynamic_date\" id=\"radio_dynamic_present_year\" value=\"present_year\" tabindex=\"0\" [(ngModel)]=\"value\">\r\n            <label class=\"custom-control-label\" for=\"radio_dynamic_present_year\">{{locale?.presentYear}}</label>\r\n        </div></li>\r\n        <li class=\"dynamic-view-item\" [class.active]=\"value === 'previou'\">\r\n            <div class=\"custom-control custom-radio\">\r\n                <input class=\"custom-control-input\" type=\"radio\" name=\"dynamic_date\" id=\"radio_dynamic_previou\" value=\"previou\" tabindex=\"0\" [(ngModel)]=\"value\">\r\n                <label class=\"custom-control-label\" for=\"radio_dynamic_previou\">\r\n                    {{locale?.previous}} <input type=\"number\" [ngModel]=\"previous_days\" name=\"previous_days\" \r\n                    (mousedown)=\"onListItemFormControlClick($event)\" (keydown)=\"onNumberKeydown($event)\"\r\n                    min=\"1\" max=\"{{previousMaxDays}}\" (input)=\"checkMaxDays($event, 'previous')\">\r\n                    <select (mousedown)=\"onListItemFormControlClick($event)\" name=\"previous_dateType\" [(ngModel)]=\"previous_dateType\">\r\n                        <option value=\"day\">{{locale?.day}}</option>\r\n                        <option value=\"week\">{{locale?.week}}</option>\r\n                        <option value=\"month\">{{locale?.month}}</option>    \r\n                        <option value=\"year\">{{locale?.year}}</option>      \r\n                    </select>\r\n                </label>\r\n            </div> </li>\r\n        <li class=\"dynamic-view-item\" [class.active]=\"value === 'future'\">\r\n            <div class=\"custom-control custom-radio\">\r\n                <input class=\"custom-control-input\" type=\"radio\" name=\"dynamic_date\" id=\"radio_dynamic_future\" value=\"future\" tabindex=\"0\" [(ngModel)]=\"value\">\r\n                <label class=\"custom-control-label\" for=\"radio_dynamic_future\">\r\n                    {{locale?.future}} <input type=\"number\" [ngModel]=\"future_days\" name=\"future_days\" \r\n                    (mousedown)=\"onListItemFormControlClick($event)\" (keydown)=\"onNumberKeydown($event)\"\r\n                    min=\"1\" max=\"{{futureMaxDays}}\" (input)=\"checkMaxDays($event, 'future')\">\r\n                    <select (mousedown)=\"onListItemFormControlClick($event)\" [(ngModel)]=\"future_dateType\" name=\"future_dateType\">\r\n                        <option value=\"day\">{{locale?.day}}</option>\r\n                        <option value=\"week\">{{locale?.week}}</option>\r\n                        <option value=\"month\">{{locale?.month}}</option>    \r\n                        <option value=\"year\">{{locale?.year}}</option>    \r\n                    </select>\r\n                </label>\r\n            </div>    \r\n        </li>\r\n        <li class=\"dynamic-view-item\" [class.active]=\"value === 'present_first'\">\r\n            <div class=\"custom-control custom-radio\">\r\n                <input class=\"custom-control-input\" type=\"radio\" name=\"dynamic_date\" id=\"radio_dynamic_present_first\" value=\"present_first\" tabindex=\"0\" [(ngModel)]=\"value\">\r\n                <label class=\"custom-control-label\" for=\"radio_dynamic_present_first\">\r\n                    {{locale?.current}}<select (mousedown)=\"onListItemFormControlClick($event)\" [(ngModel)]=\"present_first_dateType\" name=\"present_first_dateType\">\r\n                        <option value=\"week\">{{locale?.week}}</option>\r\n                        <option value=\"month\">{{locale?.month}}</option>    \r\n                        <option value=\"year\">{{locale?.year}}</option>    \r\n                    </select>{{locale?.first}}\r\n                </label>\r\n            </div>    \r\n            \r\n        </li>\r\n        <li class=\"dynamic-view-item\" [class.active]=\"value === 'present_last'\">\r\n            <div class=\"custom-control custom-radio\">\r\n                <input class=\"custom-control-input\" type=\"radio\" name=\"dynamic_date\" id=\"radio_dynamic_present_last\" value=\"present_last\" tabindex=\"0\" [(ngModel)]=\"value\">\r\n                <label class=\"custom-control-label\" for=\"radio_dynamic_present_last\" >\r\n                    {{locale?.current}}<select (mousedown)=\"onListItemFormControlClick($event)\" [(ngModel)]=\"present_last_dateType\" name=\"present_last_dateType\">\r\n                        <option value=\"week\">{{locale?.week}}</option>\r\n                        <option value=\"month\">{{locale?.month}}</option>    \r\n                        <option value=\"year\">{{locale?.year}}</option>    \r\n                    </select>{{locale?.last}}\r\n                </label>\r\n            </div>  \r\n        </li>\r\n    </ul>\r\n</div>\r\n\r\n\r\n\r\n",
                    styles: [".datepicker_dynamic-view ul{margin:10px 8px;list-style:none}.dynamic-view-item{line-height:30px;padding-left:8px;list-style:none}.dynamic-view-item.active{background:#f7f8fb;border-radius:6px}.dynamic-view-item input[type=number]{width:46px;text-align:center;-moz-appearance:textfield}.dynamic-view-item select{width:60px;padding-left:5px;margin:0 6px}.dynamic-view-item input,.dynamic-view-item select{height:26px;border-radius:6px;border:1px solid #d9dee7}.dynamic-view-item input:focus,.dynamic-view-item select:focus{outline:0}.dynamic-view-item .custom-radio .custom-control-label::before{top:7px}.dynamic-view-item label{padding-left:20px;display:inline-block;width:100%}.dynamic-view-item input::-webkit-inner-spin-button,.dynamic-view-item input::-webkit-outer-spin-button{-webkit-appearance:none!important}"]
                }] }
    ];
    /** @nocollapse */
    DynamicViewComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ChangeDetectorRef }
    ]; };
    DynamicViewComponent.propDecorators = {
        data: [{ type: Input }],
        previousMaxDays: [{ type: Input }],
        futureMaxDays: [{ type: Input }],
        locale: [{ type: Input }],
        valueChange: [{ type: Output }]
    };
    return DynamicViewComponent;
}());
export { DynamicViewComponent };
if (false) {
    /** @type {?} */
    DynamicViewComponent.prototype._value;
    /**
     * @type {?}
     * @private
     */
    DynamicViewComponent.prototype._data;
    /** @type {?} */
    DynamicViewComponent.prototype.previousMaxDays;
    /** @type {?} */
    DynamicViewComponent.prototype.futureMaxDays;
    /** @type {?} */
    DynamicViewComponent.prototype.locale;
    /** @type {?} */
    DynamicViewComponent.prototype.previous_dateType;
    /** @type {?} */
    DynamicViewComponent.prototype.previous_days;
    /** @type {?} */
    DynamicViewComponent.prototype.future_dateType;
    /** @type {?} */
    DynamicViewComponent.prototype.future_days;
    /** @type {?} */
    DynamicViewComponent.prototype.present_first_dateType;
    /** @type {?} */
    DynamicViewComponent.prototype.present_last_dateType;
    /** @type {?} */
    DynamicViewComponent.prototype.valueChange;
    /**
     * @type {?}
     * @private
     */
    DynamicViewComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    DynamicViewComponent.prototype.cd;
}
