// @farris/command-services/esm2015/lib/query-solution/query-solution.service.js
import { Injectable } from '@angular/core';
import { QueryConditionHandlerFactory } from './querycondition.handler.factory';
import { FrameContext } from '@farris/devkit';
import { FrameworkSessionService } from '@farris/bef';
export class QuerySolutionService {
    constructor(frameContext, frameworkSessionService) {
        this.frameContext = frameContext;
        this.frameworkSessionService = frameworkSessionService;
    }
    getUserSessionId() {
        return this.frameworkSessionService.getUserSessionId();
    }
    setFilterConditions(params) {
        let queryConditionHandler;
        const filterConditionList = [];
        params.forEach(qc => {
            if (!qc.value.isEmpty()) {
                queryConditionHandler = QueryConditionHandlerFactory.getInstance().getHandler(qc.control.getControlType());
                queryConditionHandler && filterConditionList.push(...queryConditionHandler.convert(qc));
            }
        });
        //设置到NgCommand params.filter属性上：{UISTATE~/root-component/filterConditionList}
        this.frameContext.uiState['filterConditionList'] = JSON.stringify(filterConditionList);
        // const originalConditions = this.getOriginalQueryConditions(filterConditionList);
        this.frameContext.uiState['originalFilterConditionList'] = JSON.stringify(filterConditionList);
    }
    setCurrentQueryConditions(conditions) {
        this.frameContext.uiState['currentQueryConditions'] = conditions;
    }
}
QuerySolutionService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
QuerySolutionService.ctorParameters = () => [
    { type: FrameContext },
    { type: FrameworkSessionService }
];
