// @edp-aif/common-api/esm5/lib/content/impl/BizEventContent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Content } from './Content';
import { BizEventEntity } from '../../entity/impl/BizEventEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
var BizEventContent = /** @class */ (function (_super) {
    tslib_1.__extends(BizEventContent, _super);
    function BizEventContent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    BizEventContent.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.canDoExpression = null;
        this.bizFlowchartIdExpression = null;
    };
    /**
     * @return {?}
     */
    BizEventContent.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "canDoExpression", this.canDoExpression);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartIdExpression", this.bizFlowchartIdExpression);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BizEventContent.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.canDoExpression = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "canDoExpression");
        this.bizFlowchartIdExpression = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartIdExpression");
    };
    /**
     * @protected
     * @return {?}
     */
    BizEventContent.prototype.GetEntityClass = /**
     * @protected
     * @return {?}
     */
    function () {
        return BizEventEntity;
    };
    return BizEventContent;
}(Content));
export { BizEventContent };
if (false) {
    /** @type {?} */
    BizEventContent.prototype.canDoExpression;
    /** @type {?} */
    BizEventContent.prototype.bizFlowchartIdExpression;
}
