// @edp-aif/common-api/esm5/lib/content/impl/BizFlowchartContent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Content } from './Content';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
import { BizFlowchartEntity } from '../../entity/impl/BizFlowchartEntity';
var BizFlowchartContent = /** @class */ (function (_super) {
    tslib_1.__extends(BizFlowchartContent, _super);
    function BizFlowchartContent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    BizFlowchartContent.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.bpmnModel = null;
    };
    /**
     * @return {?}
     */
    BizFlowchartContent.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bpmnModel", this.bpmnModel);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BizFlowchartContent.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.bpmnModel = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bpmnModel");
    };
    /**
     * @protected
     * @return {?}
     */
    BizFlowchartContent.prototype.GetEntityClass = /**
     * @protected
     * @return {?}
     */
    function () {
        return BizFlowchartEntity;
    };
    return BizFlowchartContent;
}(Content));
export { BizFlowchartContent };
if (false) {
    /** @type {?} */
    BizFlowchartContent.prototype.bpmnModel;
}
