// @edp-aif/common-api/esm5/lib/content/impl/Content.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl } from '../../JsonConvertableImpl';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * @abstract
 * @template TEntity
 */
var /**
 * @abstract
 * @template TEntity
 */
Content = /** @class */ (function (_super) {
    tslib_1.__extends(Content, _super);
    function Content() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    Content.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.entity = null;
    };
    /**
     * @return {?}
     */
    Content.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "entity", this.entity);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    Content.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.entity = BpmnModelJsonHelper.GetAttributeValue(obj, "entity");
    };
    return Content;
}(JsonConvertableImpl));
/**
 * @abstract
 * @template TEntity
 */
export { Content };
if (false) {
    /** @type {?} */
    Content.prototype.entity;
    /**
     * @abstract
     * @protected
     * @template ClassOfT
     * @return {?}
     */
    Content.prototype.GetEntityClass = function () { };
}
