// @edp-aif/common-api/esm5/lib/content/impl/CreationRuleContent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Content } from './Content';
import { CreationRuleEntity } from '../../entity/impl/CreationRuleEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
var CreationRuleContent = /** @class */ (function (_super) {
    tslib_1.__extends(CreationRuleContent, _super);
    function CreationRuleContent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    CreationRuleContent.prototype.ResetBeforeLoadFrom = /**
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.mappingContent = null;
    };
    /**
     * @return {?}
     */
    CreationRuleContent.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingContent", this.mappingContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "formSettings", this.formSettings);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CreationRuleContent.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.mappingContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingContent");
        this.formSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettings");
    };
    /**
     * @protected
     * @return {?}
     */
    CreationRuleContent.prototype.GetEntityClass = /**
     * @protected
     * @return {?}
     */
    function () {
        return CreationRuleEntity;
    };
    return CreationRuleContent;
}(Content));
export { CreationRuleContent };
if (false) {
    /** @type {?} */
    CreationRuleContent.prototype.mappingContent;
    /** @type {?} */
    CreationRuleContent.prototype.formSettings;
}
