// @edp-aif/common-api/esm5/lib/content/impl/WritebackRuleContent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Content } from './Content';
import { WritebackRuleEntity } from '../../entity/impl/WritebackRuleEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
var WritebackRuleContent = /** @class */ (function (_super) {
    tslib_1.__extends(WritebackRuleContent, _super);
    function WritebackRuleContent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    WritebackRuleContent.prototype.ResetBeforeLoadFrom = /**
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.mappingContent = null;
    };
    /**
     * @return {?}
     */
    WritebackRuleContent.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingContent", this.mappingContent);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    WritebackRuleContent.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.mappingContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingContent");
    };
    /**
     * @protected
     * @return {?}
     */
    WritebackRuleContent.prototype.GetEntityClass = /**
     * @protected
     * @return {?}
     */
    function () {
        return WritebackRuleEntity;
    };
    return WritebackRuleContent;
}(Content));
export { WritebackRuleContent };
if (false) {
    /** @type {?} */
    WritebackRuleContent.prototype.mappingContent;
}
