// @edp-bif/common-api/esm5/lib/content/impl/AbstractRtBillProcessContent.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/content/impl/AbstractRtBillProcessContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '../../ref';
import { ConditionEntity } from "../../entity/impl/ConditionEntity";
import { MappingModeEntity } from '../../entity/impl/MappingModeEntity';
import { SaveModeEntity } from '../../entity/impl/SaveModeEntity';
/**
 * 单据流内容类。
 * @param TEntity 单据流实体。
 * @abstract
 * @template TEntity
 */
var /**
 * 单据流内容类。
 * @param TEntity 单据流实体。
 * @abstract
 * @template TEntity
 */
AbstractRtBillProcessContent = /** @class */ (function (_super) {
    tslib_1.__extends(AbstractRtBillProcessContent, _super);
    function AbstractRtBillProcessContent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    AbstractRtBillProcessContent.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.entity = null;
        this.eventRouterConditionsContent = new Array();
        this.mappingConditionsContent = new Array();
        this.mappingModeContent = null;
        this.saveModeContent = null;
    };
    /**
     * @return {?}
     */
    AbstractRtBillProcessContent.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "entity", this.entity);
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsContent", this.eventRouterConditionsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsContent", this.mappingConditionsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeContent", this.mappingModeContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeContent", this.saveModeContent);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AbstractRtBillProcessContent.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.entity = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "entity", this.getEntityClass());
        this.eventRouterConditionsContent = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "eventRouterConditionsContent", ConditionEntity);
        this.mappingConditionsContent = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingConditionsContent", ConditionEntity);
        this.mappingModeContent = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "mappingModeContent", MappingModeEntity);
        this.saveModeContent = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "saveModeContent", SaveModeEntity);
    };
    return AbstractRtBillProcessContent;
}(JsonConvertableImpl));
/**
 * 单据流内容类。
 * @param TEntity 单据流实体。
 * @abstract
 * @template TEntity
 */
export { AbstractRtBillProcessContent };
if (false) {
    /** @type {?} */
    AbstractRtBillProcessContent.prototype.entity;
    /** @type {?} */
    AbstractRtBillProcessContent.prototype.eventRouterConditionsContent;
    /** @type {?} */
    AbstractRtBillProcessContent.prototype.mappingConditionsContent;
    /** @type {?} */
    AbstractRtBillProcessContent.prototype.mappingModeContent;
    /** @type {?} */
    AbstractRtBillProcessContent.prototype.saveModeContent;
    /**
     * 获取泛型类类型。
     *
     * @abstract
     * @protected
     * @template ClassOfT
     * @return {?} 泛型类类型。
     */
    AbstractRtBillProcessContent.prototype.getEntityClass = function () { };
}
