// @ecp-caf/caf-common/esm2015/lib/utilities/common.utilities.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class CommonUtil {
    /**
     * @param {?} paraName
     * @return {?}
     */
    static getQueryString(paraName) {
        /** @type {?} */
        const reg = new RegExp('(^|&)' + paraName + '=([^&]*)(&|$)', 'i');
        /** @type {?} */
        const r = window.location.search.substr(1).match(reg);
        /** @type {?} */
        let paraValue = '';
        if (r) {
            paraValue = r[2];
        }
        return paraValue;
    }
    /**
     * @param {?} paraName
     * @return {?}
     */
    static getQueryStringFromHash(paraName) {
        /** @type {?} */
        const url = window.location.hash;
        /** @type {?} */
        const arrObj = url && url.split('?');
        if (arrObj && arrObj.length > 1) {
            /** @type {?} */
            const arrPara = arrObj[1].split('&');
            if (!arrPara) {
                return '';
            }
            /** @type {?} */
            let arr;
            // tslint:disable-next-line: prefer-for-of
            for (let i = 0; i < arrPara.length; i++) {
                arr = arrPara[i].split('=');
                if (arr != null && arr[0] === paraName) {
                    return arr[1];
                }
            }
            return '';
        }
        else {
            return '';
        }
    }
    /**
     * @return {?}
     */
    static getCookieMapFromDoc() {
        /** @type {?} */
        const cookieArr = document.cookie && document.cookie.split(';');
        /** @type {?} */
        const cookieMap = new Map();
        if (cookieArr) {
            cookieArr.forEach((/**
             * @param {?} item
             * @return {?}
             */
            item => {
                /** @type {?} */
                const params = item.split('=');
                cookieMap.set(params[0], params[1]);
            }));
        }
        return cookieMap;
    }
    /**
     * 获取当前window 中funcId appId等公共变量
     * 返回结果格式：funcId=xx&appType=xxx&
     * @return {?}
     */
    static getCommonVariables() {
        return CommonUtil.getQueryStringFromHash('cvft');
    }
    /**
     * @private
     * @param {?} item
     * @return {?}
     */
    static matchCommonVariableKey(item) {
        /** @type {?} */
        const itemArr = item.split('=');
        if (!itemArr || itemArr.length !== 2 || !itemArr[0]) {
            return false;
        }
        switch (itemArr[0].toLowerCase()) {
            case 'funcid': return true;
            case 'appid': return true;
            case 'appentrance': return true;
            case 'apptype': return true;
        }
        return false;
    }
}
