// @farris/mobile-ui/lib/switch/index.js
import { computed, openBlock, createBlock, createVNode, createCommentVNode } from 'vue';

var script = {
  name: 'fm-switch',
  props: {
    modelValue: {
      type: Boolean,
      "default": false
    },
    disabled: {
      type: Boolean,
      "default": false
    },
    readonly: {
      type: Boolean,
      "default": false
    },
    loading: {
      type: Boolean,
      "default": false
    },
    size: {
      type: Number,
      "default": 28
    },
    activeColor: {
      type: String,
      "default": ''
    },
    inactiveColor: {
      type: String,
      "default": ''
    },
    activeValue: {
      type: [Boolean, Number, String],
      "default": true
    },
    inactiveValue: {
      type: [Boolean, Number, String],
      "default": false
    }
  },
  setup: function setup(props, _ref) {
    var emit = _ref.emit;
    //   const innerValue = ref(props.modelValue)
    var checked = computed(function () {
      return props.modelValue === props.activeValue;
    });
    var style = computed(function () {
      return {
        fontSize: props.size + 'px',
        backgroundColor: checked ? props.activeColor : props.inactiveColor
      };
    });

    var onClick = function onClick(event) {
      emit('click', event);

      if (!props.disabled && !props.loading && !props.readonly) {
        var innerValue = checked.value ? props.inactiveValue : props.activeValue;
        emit('update:modelValue', innerValue);
        emit('change', innerValue);
      }
    };

    return {
      style: style,
      checked: checked,
      onClick: onClick
    };
  }
};

var _hoisted_1 = {
  "class": "fm-switch-node"
};
var _hoisted_2 = {
  key: 0,
  "class": "fm-switch-loading-icon"
};

var _hoisted_3 = /*#__PURE__*/createVNode("svg", {
  viewBox: "25 25 50 50",
  "class": "fm-switch-loading-icon-circular"
}, [/*#__PURE__*/createVNode("circle", {
  cx: "50",
  cy: "50",
  r: "20",
  fill: "none"
})], -1
/* HOISTED */
);

function render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createBlock("div", {
    "class": ["fm-switch", {
      'fm-switch-on': $setup.checked,
      'fm-switch-loading': $props.loading,
      'fm-switch-disabled': $props.disabled,
      'fm-switch-readonly': $props.readonly
    }],
    style: $setup.style,
    onClick: _cache[1] || (_cache[1] = function () {
      return $setup.onClick && $setup.onClick.apply($setup, arguments);
    })
  }, [createVNode("div", _hoisted_1, [$props.loading ? (openBlock(), createBlock("span", _hoisted_2, [_hoisted_3])) : createCommentVNode("v-if", true)])], 6
  /* CLASS, STYLE */
  );
}

script.render = render;
script.__file = "packages/switch/src/index.vue";

script.install = function (app) {
  app.component(script.name, script);
};

export default script;
export { script as Switch };
