// @gsp-wf/wf-process-editor/esm5/lib/views/jump-scope/jump-scope.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, HostBinding, Output, EventEmitter, Input, ViewChild, TemplateRef, Optional, ChangeDetectorRef, LOCALE_ID, Inject } from '@angular/core';
import { ProcessDesignerUIState } from '../../app/process-designer.uistate';
import { DialogComponent } from '@farris/ui-dialog';
import { MessagerService } from '@farris/ui-messager';
import { DatagridComponent } from '@farris/ui-datagrid';
import { LookupGridComponent } from '@farris/ui-lookup';
import { of } from 'rxjs';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../i18n/pipe/wf-process-editor-locale.pipe';
var JumpScopeComponent = /** @class */ (function () {
    function JumpScopeComponent(uistate, message, cdr, localeId, resources, localePipe) {
        var _this = this;
        this.uistate = uistate;
        this.message = message;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.cls = 'farris-main-area flex-column';
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.modalConfig = {
            title: this.localePipe.transform('component.jumpScope.targetNodeOptions'),
            width: 520,
            height: 460,
            showButtons: true
        };
        this.cols = [
            { title: this.localePipe.transform('component.jumpScope.name'), field: 'Name', width: 60, searchField: 'Name' },
            { title: 'ID', field: 'Id', width: 90, searchField: 'Id' }
        ];
        // ==========================================================
        this.actis = [];
        this.fullActis = [];
        this.defaultacty = '';
        this.categoryListItems = [
            { value: 'OutConnectNode', name: this.localePipe.transform('component.jumpScope.outConnectNode') },
            { value: 'Global', name: this.localePipe.transform('component.jumpScope.global') },
            { value: 'AfterNodes', name: this.localePipe.transform('component.jumpScope.afterNodes') },
            { value: 'SpecifiedSet', name: this.localePipe.transform('component.jumpScope.specifiedSet') }
        ];
        this.category = '';
        this.defaultType = '';
        this.defaultTypeListItems = [
            { value: 'FirstNode', name: this.localePipe.transform('component.jumpScope.firstNode') },
            { value: 'NullNode', name: this.localePipe.transform('component.jumpScope.nullNode') },
            { value: 'SpecifyNode', name: this.localePipe.transform('component.jumpScope.specifyNode') }
        ];
        this.defaultActivity = '';
        this.pageInfo = {
            pageIndex: 1,
            pageList: [10, 20, 30, 50],
            pageSize: 20
        };
        this.actisSelect = (/**
         * @param {?} rowData
         * @return {?}
         */
        function (rowData) {
            // 附加节点的帮助后事件，把选中的数据赋给fullActis（此处处理的目的是为了显示Name同时取Id值）
            // console.log(rowData);
            _this.fullActis = rowData;
            return of({
                closeDialog: true
            });
        });
        this.defaultActivitySelect = (/**
         * @param {?} rowData
         * @return {?}
         */
        function (rowData) {
            // 默认节点的帮助后事件，把选中的id赋给defaultActivity（此处处理的目的是为了显示Name同时取Id值）
            _this.defaultActivity = rowData.Id;
            return of({
                closeDialog: true
            });
        });
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
    }
    Object.defineProperty(JumpScopeComponent.prototype, "value", {
        set: /**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            var e_1, _a, e_2, _b;
            if (data) {
                this.category = data.Category;
                this.defaultType = data.DefaultType;
                // 取到DefaultActivity的节点Id，遍历出Name绑定到radiogroup上去（defaultacty）
                /** @type {?} */
                var acty = data.DefaultActivity;
                if (acty != null) {
                    try {
                        for (var _c = tslib_1.__values(this.uistate.allActList), _d = _c.next(); !_d.done; _d = _c.next()) {
                            var act = _d.value;
                            if (act.Id === acty) {
                                this.defaultacty = act.Name;
                            }
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                }
                // 取到DestinationActivities的节点Id，遍历出Id和Name绑定到radiogroup上去（actis）
                /** @type {?} */
                var strs = data.DestinationActivities;
                if (strs && strs.length > 0) {
                    var _loop_1 = function (str) {
                        /** @type {?} */
                        var name_1 = this_1.uistate.allActList.filter((/**
                         * @param {?} act
                         * @return {?}
                         */
                        function (act) { return act.Id === str; }))[0].Name;
                        // const acti = { Id: str, Name: name };
                        this_1.actis.push(name_1);
                    };
                    var this_1 = this;
                    try {
                        for (var strs_1 = tslib_1.__values(strs), strs_1_1 = strs_1.next(); !strs_1_1.done; strs_1_1 = strs_1.next()) {
                            var str = strs_1_1.value;
                            _loop_1(str);
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (strs_1_1 && !strs_1_1.done && (_b = strs_1.return)) _b.call(strs_1);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                }
            }
            else {
                this.category = 'OutConnectNode';
                this.defaultType = 'FirstNode';
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    JumpScopeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    JumpScopeComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        var e_3, _a, e_4, _b, e_5, _c;
        if (this.validator()) {
            /** @type {?} */
            var value = void 0;
            /** @type {?} */
            var list_1 = [];
            // 重新打开设置界面，没有进行节点配置的情况（DefaultActivity）
            if (this.defaultActivity === '' && this.defaultacty !== '') {
                try {
                    for (var _d = tslib_1.__values(this.uistate.allActList), _e = _d.next(); !_e.done; _e = _d.next()) {
                        var defAct = _e.value;
                        if (defAct.Name === this.defaultacty) {
                            this.defaultActivity = defAct.Id;
                        }
                    }
                }
                catch (e_3_1) { e_3 = { error: e_3_1 }; }
                finally {
                    try {
                        if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
                    }
                    finally { if (e_3) throw e_3.error; }
                }
            }
            // 重新打开设置界面，没有进行节点配置的情况（DestinationActivities）
            if (this.fullActis.length === 0 && this.actis.length > 0) {
                try {
                    for (var _f = tslib_1.__values(this.actis), _g = _f.next(); !_g.done; _g = _f.next()) {
                        var act = _g.value;
                        try {
                            for (var _h = tslib_1.__values(this.uistate.allActList), _j = _h.next(); !_j.done; _j = _h.next()) {
                                var allAct = _j.value;
                                if (act === allAct.Name) {
                                    list_1.push(allAct.Id);
                                }
                            }
                        }
                        catch (e_5_1) { e_5 = { error: e_5_1 }; }
                        finally {
                            try {
                                if (_j && !_j.done && (_c = _h.return)) _c.call(_h);
                            }
                            finally { if (e_5) throw e_5.error; }
                        }
                    }
                }
                catch (e_4_1) { e_4 = { error: e_4_1 }; }
                finally {
                    try {
                        if (_g && !_g.done && (_b = _f.return)) _b.call(_f);
                    }
                    finally { if (e_4) throw e_4.error; }
                }
            }
            else if (this.category === 'SpecifiedSet' && this.fullActis && this.fullActis.length > 0) {
                // 附加节点的id传给模型上的DestinationActivities
                this.fullActis.forEach((/**
                 * @param {?} act
                 * @return {?}
                 */
                function (act) { return list_1.push(act.Id); }));
            }
            value = {
                Category: this.category,
                DefaultType: this.defaultType,
                DefaultActivity: this.defaultActivity,
                DestinationActivities: list_1,
            };
            /** @type {?} */
            var object = { value: value, parameters: null };
            this.submitModal.emit(object);
        }
    };
    /**
     * @return {?}
     */
    JumpScopeComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeModal.emit();
    };
    /**
     * @private
     * @return {?}
     */
    JumpScopeComponent.prototype.validator = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.category) {
            this.message.warning(this.localePipe.transform('component.jumpScope.noSetTargetNode'));
            return false;
        }
        // if (this.category === 'SpecifiedSet' && Object.prototype.isPrototypeOf(this.actis)) {
        //   this.message.info('尚未选择附加节点，请了解！');
        //   return false;
        // }
        if (this.defaultType === 'SpecifyNode' && this.defaultacty === '') {
            this.message.warning(this.localePipe.transform('component.jumpScope.noSetSpecifyNode'));
            return false;
        }
        return true;
    };
    /**
     * @param {?} arg
     * @return {?}
     */
    JumpScopeComponent.prototype.searchDefault = /**
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        this.pageInfo.pageIndex = arg.pageInfo.pageIndex;
        this.pageInfo.pageSize = arg.pageInfo.pageSize;
        /** @type {?} */
        var search = arg.search;
        this.loadDefault(search);
    };
    /**
     * @param {?=} search
     * @return {?}
     */
    JumpScopeComponent.prototype.loadDefault = /**
     * @param {?=} search
     * @return {?}
     */
    function (search) {
        /** @type {?} */
        var start = (this.pageInfo.pageIndex - 1) * this.pageInfo.pageSize;
        /** @type {?} */
        var end = start + this.pageInfo.pageSize;
        /** @type {?} */
        var items = [];
        if (search && search.value) {
            if (search.field === '*') {
                items = this.uistate.allActList.filter((/**
                 * @param {?} i
                 * @return {?}
                 */
                function (i) { return i.Name.indexOf(search.value) > -1 || i.Id.indexOf(search.value) > -1; }));
            }
            else {
                items = this.uistate.allActList.filter((/**
                 * @param {?} f
                 * @return {?}
                 */
                function (f) { return f[search.field].indexOf(search.value) > -1; }));
            }
        }
        else {
            items = this.uistate.allActList;
        }
        this.total = items.length;
        // this.items = items;
        if (items) {
            this.allActisTable.componentRef.instance.loadData({
                data: items
            });
        }
    };
    /**
     * @param {?} arg
     * @return {?}
     */
    JumpScopeComponent.prototype.searchSelect = /**
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        this.pageInfo.pageIndex = arg.pageInfo.pageIndex;
        this.pageInfo.pageSize = arg.pageInfo.pageSize;
        /** @type {?} */
        var search = arg.search;
        this.loadSelect(search);
    };
    /**
     * @param {?=} search
     * @return {?}
     */
    JumpScopeComponent.prototype.loadSelect = /**
     * @param {?=} search
     * @return {?}
     */
    function (search) {
        /** @type {?} */
        var start = (this.pageInfo.pageIndex - 1) * this.pageInfo.pageSize;
        /** @type {?} */
        var end = start + this.pageInfo.pageSize;
        /** @type {?} */
        var items = [];
        if (search && search.value) {
            if (search.field === '*') {
                items = this.uistate.allActList.filter((/**
                 * @param {?} i
                 * @return {?}
                 */
                function (i) { return i.Name.indexOf(search.value) > -1 || i.Id.indexOf(search.value) > -1; }));
            }
            else {
                items = this.uistate.allActList.filter((/**
                 * @param {?} f
                 * @return {?}
                 */
                function (f) { return f[search.field].indexOf(search.value) > -1; }));
            }
        }
        else {
            items = this.uistate.allActList;
        }
        this.total = items.length;
        // this.items = items;
        if (items) {
            this.userActis.componentRef.instance.loadData({
                data: items
            });
        }
    };
    /**
     * @return {?}
     */
    JumpScopeComponent.prototype.deleteRow = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.actTable.selectedRow && this.actTable.selectedRow.data) {
            /** @type {?} */
            var i = this.actis.findIndex((/**
             * @param {?} d
             * @return {?}
             */
            function (d) { return d.id === _this.actTable.selectedRow.data.id; }));
            this.actis.splice(i, 1);
            this.actTable.loadData(this.actis);
        }
        else {
            this.message.info(this.localePipe.transform('component.jumpScope.selectedRow'));
        }
    };
    JumpScopeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-jump-scope',
                    template: "<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <div class=\"f-utils-fill-flex-column h-100\">\r\n        <form class=\"h-100\">\r\n          <div class=\"f-utils-fill-flex-column h-100\">\r\n            <!-- \u5141\u8BB8\u9009\u62E9\u76EE\u6807\u8282\u70B9\u5355\u9009\u9879 -->\r\n            <!-- <div class=\"mx-4 mt-3 farris-form-controls-inline\">\r\n              <div class=\"farris-group-wrap\">\r\n                <div class=\"form-group farris-form-group\">\r\n                  <div class=\"farris-input-wrap\">\r\n                    <div class=\"custom-control custom-checkbox\">\r\n                      <input class=\"custom-control-input\" name=\"submitterAutoPass\" type=\"checkbox\"\r\n                        id=\"submitterAutoPass\" [(ngModel)]=\"outConnectNode\" />\r\n                      <label class=\"custom-control-label\" for=\"submitterAutoPass\">\u5141\u8BB8\u9009\u62E9\u76EE\u6807\u8282\u70B9</label>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div> -->\r\n            <!-- \u53EF\u9009\u8282\u70B9\u5C0F\u6807\u9898 -->\r\n            <div class=\"mx-2 mt-1\">\r\n              <div class=\"title-content\">\r\n                <h4 class=\"title-text\">{{'component.jumpScope.optionalNodes'|wfLocale}}</h4><span class=\"title-line\"></span>\r\n              </div>\r\n            </div>\r\n            <div class=\"mt-2 mx-4 farris-form-controls-inline\">\r\n              <div class=\"farris-group-wrap\">\r\n                <div class=\"form-group\">\r\n                  <!-- \u7AD6\u5217\u679A\u4E3E\u7684category\u9009\u9879 -->\r\n                  <farris-radiogroup [(ngModel)]=\"category\" [data]=\"categoryListItems\" name=\"category\"\r\n                    [horizontal]=false>\r\n                  </farris-radiogroup>\r\n                </div>\r\n                <!-- \u9644\u52A0\u8282\u70B9\u5E2E\u52A9 -->\r\n                <farris-lookup-grid #userActisDialog\r\n                  [items]=\"uistate.allActList | jumpScopePipe:uistate.propertyData.Id \" [columns]=\"cols\"\r\n                  [disabled]=\"!(category === 'SpecifiedSet')\" [resizable]=\"true\" [singleSelect]=\"false\" [idField]=\"'Id'\"\r\n                  [textField]=\"'Name'\" [(ngModel)]=\"actis\" [valueField]=\"'Id'\" [pagination]=\"true\"\r\n                  [pageIndex]=\"pageInfo.pageIndex\" name=\"defaultActivities\" [pageSize]=\"pageInfo.pageSize\" (search)=\"searchSelect($event)\"\r\n                  [pageList]=\"pageInfo.pageList\" [total]=\"total\" [enableClear]=\"false\" title=\"{{'component.jumpScope.listOfAdditional'|wfLocale}}\"\r\n                  [displayType]=\"'List'\" [remoteSearch]='false' [enableToSelect]=\"false\" placeholder=\"{{'component.jumpScope.pleaseSelect1'|wfLocale}}\" [dictPicked]=\"actisSelect\">\r\n                </farris-lookup-grid>\r\n              </div>\r\n            </div>\r\n            <!-- \u9ED8\u8BA4\u8282\u70B9\u5C0F\u6807\u9898 -->\r\n            <div class=\"mx-2 mt-1\">\r\n              <div class=\"title-content\">\r\n                <h4 class=\"title-text\">{{'component.jumpScope.defaultNode'|wfLocale}}</h4><span class=\"title-line\"></span>\r\n              </div>\r\n            </div>\r\n            <div class=\"mt-2 mx-4 farris-form-controls-inline\">\r\n              <div class=\"farris-group-wrap\">\r\n                <div class=\"form-group\">\r\n                  <!-- \u7AD6\u5217\u679A\u4E3E\u7684DefaultType\u9009\u9879 -->\r\n                  <farris-radiogroup [(ngModel)]=\"defaultType\" [data]=\"defaultTypeListItems\" name=\"defaultType\"\r\n                    [horizontal]=false>\r\n                  </farris-radiogroup>\r\n                </div>\r\n                <!-- \u9ED8\u8BA4\u8282\u70B9\u5E2E\u52A9 -->\r\n                <farris-lookup-grid #allActisTable [items]=\"uistate.allActList | jumpScopePipe:uistate.propertyData.Id \"\r\n                  [columns]=\"cols\" [disabled]=\"!(defaultType === 'SpecifyNode')\" [resizable]=\"true\"\r\n                  [singleSelect]=\"true\" [idField]=\"'Id'\" [textField]=\"'Name'\" [(ngModel)]=\"defaultacty\"\r\n                  [valueField]=\"'Id'\" [pagination]=\"true\" placeholder=\"{{'component.jumpScope.pleaseSelect2'|wfLocale}}\" [pageIndex]=\"pageInfo.pageIndex\"\r\n                  name=\"defaultActivity\" [pageSize]=\"pageInfo.pageSize\" [pageList]=\"pageInfo.pageList\" [total]=\"total\"\r\n                  [enableClear]=\"false\" title=\"{{'component.jumpScope.defaultNodeList'|wfLocale}}\" (search)=\"searchDefault($event)\" [displayType]=\"'List'\" [remoteSearch]='false' [dictPicked]=\"defaultActivitySelect\">\r\n                </farris-lookup-grid>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </form>\r\n      </div>\r\n    </section>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm\" (click)=\"clickCancel()\">{{'component.jumpScope.clickCancel'|wfLocale}}</button>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">{{'component.jumpScope.clickOK'|wfLocale}}</button>\r\n</ng-template>\r\n\r\n<!-- <ng-template #buttonRef>\r\n  <button class=\"btn btn-sm btn-primary mx-2\" (click)=\"actisSelect()\">\u786E\u5B9A</button>\r\n  <button class=\"btn btn-sm btn-default\" (click)=\"allActisDialog.close()\">\u53D6\u6D88</button>\r\n</ng-template> -->\r\n<!-- <farris-dialog #allActisDialog [buttons]=\"buttonRef\" [width]=\"700\" [height]=\"520\" title=\"'\u6D3B\u52A8\u5217\u8868'\">\r\n  <div class=\"f-utils-fill-flex-column farris-overflow-y-auto h-100 m-2\">\r\n    <farris-datagrid #allActisTable [columns]=\"cols\"\r\n      [data]=\"uistate.userActList | jumpScopePipe:uistate.propertyData.Id\" [idField]=\"'Id'\" [fit]=\"true\"\r\n      [fitColumns]=\"true\" [pagination]='false' [multiSelect]=\"true\" [showCheckbox]=\"true\" [multiSelect]=\"true\"\r\n      [checkOnSelect]=\"true\" [selectOnCheck]=\"true\">\r\n    </farris-datagrid>\r\n  </div>\r\n</farris-dialog> -->\r\n",
                    providers: [
                        WfLocalePipe,
                    ],
                    styles: [".title-content{color:#000;display:inline-flex;align-items:center;line-height:1.375rem;position:relative;overflow:hidden;padding:0 0 0 1rem;min-width:760px}.title-content ::before{content:'';width:.25rem;height:1.125rem;background:#59a1ff;border-radius:1px;position:absolute;top:50%;left:0;margin-top:-.5625rem}.title-text{font-size:1rem;line-height:1.375rem;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;margin-bottom:0;margin-right:10px}.title-line{flex-grow:1;border-bottom:1px dashed rgba(0,0,0,.1)}"]
                }] }
    ];
    /** @nocollapse */
    JumpScopeComponent.ctorParameters = function () { return [
        { type: ProcessDesignerUIState },
        { type: MessagerService },
        { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
        { type: WfLocalePipe, decorators: [{ type: Optional }] }
    ]; };
    JumpScopeComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        closeModal: [{ type: Output }],
        submitModal: [{ type: Output }],
        value: [{ type: Input }],
        modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }],
        allActisTable: [{ type: ViewChild, args: ['allActisTable',] }],
        allActisDialog: [{ type: ViewChild, args: ['allActisDialog',] }],
        actTable: [{ type: ViewChild, args: ['actTable',] }],
        lookup: [{ type: ViewChild, args: ['lookup',] }],
        userActis: [{ type: ViewChild, args: ['userActisDialog',] }]
    };
    return JumpScopeComponent;
}());
export { JumpScopeComponent };
if (false) {
    /** @type {?} */
    JumpScopeComponent.prototype.cls;
    /** @type {?} */
    JumpScopeComponent.prototype.closeModal;
    /** @type {?} */
    JumpScopeComponent.prototype.submitModal;
    /** @type {?} */
    JumpScopeComponent.prototype.modalFooter;
    /** @type {?} */
    JumpScopeComponent.prototype.modalConfig;
    /** @type {?} */
    JumpScopeComponent.prototype.allActisTable;
    /** @type {?} */
    JumpScopeComponent.prototype.allActisDialog;
    /** @type {?} */
    JumpScopeComponent.prototype.actTable;
    /** @type {?} */
    JumpScopeComponent.prototype.lookup;
    /** @type {?} */
    JumpScopeComponent.prototype.userActis;
    /** @type {?} */
    JumpScopeComponent.prototype.cols;
    /** @type {?} */
    JumpScopeComponent.prototype.actis;
    /** @type {?} */
    JumpScopeComponent.prototype.fullActis;
    /** @type {?} */
    JumpScopeComponent.prototype.defaultacty;
    /** @type {?} */
    JumpScopeComponent.prototype.categoryListItems;
    /** @type {?} */
    JumpScopeComponent.prototype.category;
    /** @type {?} */
    JumpScopeComponent.prototype.defaultType;
    /** @type {?} */
    JumpScopeComponent.prototype.defaultTypeListItems;
    /** @type {?} */
    JumpScopeComponent.prototype.defaultActivity;
    /** @type {?} */
    JumpScopeComponent.prototype.pageInfo;
    /** @type {?} */
    JumpScopeComponent.prototype.outConnectNode;
    /** @type {?} */
    JumpScopeComponent.prototype.total;
    /** @type {?} */
    JumpScopeComponent.prototype.actisSelect;
    /** @type {?} */
    JumpScopeComponent.prototype.defaultActivitySelect;
    /** @type {?} */
    JumpScopeComponent.prototype.uistate;
    /**
     * @type {?}
     * @private
     */
    JumpScopeComponent.prototype.message;
    /**
     * @type {?}
     * @private
     */
    JumpScopeComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    JumpScopeComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    JumpScopeComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    JumpScopeComponent.prototype.localePipe;
}
var JumpScopeConverter = /** @class */ (function () {
    function JumpScopeConverter(localePipe) {
        this.localePipe = localePipe;
    }
    /**
     * @return {?}
     */
    JumpScopeConverter.prototype.convertTo = /**
     * @return {?}
     */
    function () {
        return this.localePipe.transform('component.jumpScope.set');
    };
    return JumpScopeConverter;
}());
export { JumpScopeConverter };
if (false) {
    /**
     * @type {?}
     * @private
     */
    JumpScopeConverter.prototype.localePipe;
}
