// @gsp-wf/wf-process-editor/esm5/lib/views/shared/event-bus.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { Subject } from 'rxjs/Subject';
import { map, filter } from 'rxjs/operators';
/**
 * @record
 */
export function EventBusMessage() { }
if (false) {
    /** @type {?} */
    EventBusMessage.prototype.channel;
    /** @type {?} */
    EventBusMessage.prototype.payload;
}
var EventBusService1 = /** @class */ (function () {
    function EventBusService1() {
        var _this = this;
        this.listen = new Observable((/**
         * @param {?} observer
         * @return {?}
         */
        function (observer) {
            _this.next = (/**
             * @param {?} payload
             * @return {?}
             */
            function (payload) { return observer.next(payload); });
            return (/**
             * @param {?} _
             * @return {?}
             */
            function (_) { return _; } /* nothing to clean */);
        }));
        this.channelEmmiter = this.channelEmmiter.bind(this);
        this.emit = this.emit.bind(this);
    }
    /**
     * @return {?}
     */
    EventBusService1.prototype.allChannels = /**
     * @return {?}
     */
    function () {
        return this.listen;
    };
    /**
     * @param {?} channelName
     * @return {?}
     */
    EventBusService1.prototype.channel = /**
     * @param {?} channelName
     * @return {?}
     */
    function (channelName) {
        return this.listen
            .pipe(filter((/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return e.channel === channelName; })), map((/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return e.payload; })));
    };
    /**
     * @param {?} channelName
     * @return {?}
     */
    EventBusService1.prototype.channelEmmiter = /**
     * @param {?} channelName
     * @return {?}
     */
    function (channelName) {
        var _this = this;
        console.log('made listener for', channelName);
        return (/**
         * @param {?} payload
         * @return {?}
         */
        function (payload) {
            _this.emit({ channel: channelName, payload: payload });
        });
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    EventBusService1.prototype.emit = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var channel = _a.channel, payload = _a.payload;
        console.log('emit', channel);
        this.next({ channel: channel, payload: payload });
    };
    EventBusService1.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    EventBusService1.ctorParameters = function () { return []; };
    return EventBusService1;
}());
export { EventBusService1 };
if (false) {
    /** @type {?} */
    EventBusService1.prototype.listen;
    /**
     * @type {?}
     * @private
     */
    EventBusService1.prototype.next;
}
/**
 * Global EventEmitter
 * @template Events
 */
var EventBusService = /** @class */ (function () {
    function EventBusService() {
        /**
         * Map event name to Subject
         */
        this._subjects = new Map();
    }
    /**
     *
     * @param eventName
     * @param handler
     * @returns
     */
    /**
     *
     * @param {?} eventName
     * @param {?} handler
     * @return {?}
     */
    EventBusService.prototype.on = /**
     *
     * @param {?} eventName
     * @param {?} handler
     * @return {?}
     */
    function (eventName, handler) {
        var _this = this;
        if (!this._subjects.has(eventName)) {
            this._subjects.set(eventName, new Subject());
        }
        /** @type {?} */
        var _subject = this._subjects.get(eventName);
        /** @type {?} */
        var subscription = _subject.subscribe(handler);
        return {
            unsubscribe: (/**
             * @return {?}
             */
            function () {
                subscription.unsubscribe();
                if (_subject.observers.length > 0) {
                    _this._subjects.delete(eventName);
                }
            })
        };
    };
    /**
     *
     * @param eventName
     * @param data
     */
    /**
     *
     * @param {?} eventName
     * @param {?} data
     * @return {?}
     */
    EventBusService.prototype.emit = /**
     *
     * @param {?} eventName
     * @param {?} data
     * @return {?}
     */
    function (eventName, data) {
        if (this._subjects.has(eventName)) {
            this._subjects.get(eventName).next(data);
        }
    };
    /**
     * Destory all subscription and clear the map
     */
    /**
     * Destory all subscription and clear the map
     * @return {?}
     */
    EventBusService.prototype.destroy = /**
     * Destory all subscription and clear the map
     * @return {?}
     */
    function () {
        this._subjects.forEach((/**
         * @param {?} subject
         * @param {?} _
         * @return {?}
         */
        function (subject, _) {
            subject.unsubscribe();
        }));
        this._subjects.clear();
    };
    EventBusService.decorators = [
        { type: Injectable }
    ];
    return EventBusService;
}());
export { EventBusService };
if (false) {
    /**
     * Map event name to Subject
     * @type {?}
     * @private
     */
    EventBusService.prototype._subjects;
}
