// @farris/ui-datagrid-editors/esm2015/lib/editor-default-options.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ShowType } from '@farris/ui-datepicker';
/**
 * @record
 */
export function InputOptions() { }
if (false) {
    /** @type {?|undefined} */
    InputOptions.prototype.placeholder;
    /** @type {?|undefined} */
    InputOptions.prototype.isPassword;
    /** @type {?|undefined} */
    InputOptions.prototype.minLength;
    /** @type {?|undefined} */
    InputOptions.prototype.maxLength;
    /* Skipping unhandled member: [key: string]: any;*/
}
/** @type {?} */
export const InputDefaultOptions = {
    placeholder: '',
    isPassword: false,
    minLength: undefined,
    maxLength: undefined
};
/** @type {?} */
export const TextareaDefaultOptions = {
    editType: 'default',
    // 'dialog'
    dialogWidth: 400,
    dialogHeight: 300,
    enableWordCount: false,
    countType: 'length',
    onlyShowInDialog: true
};
/** @type {?} */
export const LookupDefaultOptions = {
    dialogWidth: 500,
    textAlign: 'left',
    dialogHeight: 600,
    showMaxButton: true,
    showCloseButton: true,
    resizable: true,
    buttonAlign: 'center',
    searchOnServer: true,
    nosearch: false,
    editable: false,
    expandLevel: -1,
    readonly: false,
    useTip: false,
    useFavorite: false,
    isRecordSize: false,
    loadTreeDataType: 'default',
    enableFullTree: false,
    enableCascade: false,
    cascadeStatus: 'enable',
    enableClear: true,
    useExtendInfo: false,
    extInfoFields: '',
    extInfoFormatter: null,
    customFormatter: null,
    customNavFormatter: null,
    selectFirstInNav: false,
    loadDataWhenOpen: true,
    treeInfo: null,
    treeTableOptions: null,
    showCheckAll: false,
    quickSelect: {
        enable: false,
        showMore: true,
        showItemsCount: 10,
        footerHeight: 0
    },
    viewType: 'text',
    showCascadeControl: true,
    showNavigation: true,
    showSelected: false,
    beforeLoadData: null,
    beforeSelectData: null,
    allowQueryFields: null,
    useNewLayout: false,
    enableMultiFieldSearch: false,
    labels: null,
    searchBarMode: 'both',
    multipleChoiceSeparator: ','
};
/**
 * @record
 */
export function InputGroupOptions() { }
if (false) {
    /** @type {?} */
    InputGroupOptions.prototype.showClear;
    /** @type {?} */
    InputGroupOptions.prototype.groupText;
    /** @type {?} */
    InputGroupOptions.prototype.disabled;
    /** @type {?} */
    InputGroupOptions.prototype.readonly;
    /** @type {?} */
    InputGroupOptions.prototype.editable;
    /** @type {?|undefined} */
    InputGroupOptions.prototype.enableViewPassword;
    /** @type {?|undefined} */
    InputGroupOptions.prototype.textAlign;
}
/** @type {?} */
export const InputGroupDefaultOptions = {
    showClear: true,
    groupText: '',
    disabled: false,
    readonly: false,
    editable: true,
    isPassword: false,
    enableViewPassword: true,
    textAlign: 'left',
    minLength: undefined,
    maxLength: undefined
};
/**
 * @record
 */
export function DatePickerOptions() { }
if (false) {
    /** @type {?} */
    DatePickerOptions.prototype.disabled;
    /** @type {?} */
    DatePickerOptions.prototype.readonly;
    /** @type {?} */
    DatePickerOptions.prototype.editable;
    /** @type {?} */
    DatePickerOptions.prototype.locale;
    /** @type {?} */
    DatePickerOptions.prototype.dateRange;
    /** @type {?} */
    DatePickerOptions.prototype.dateRangeDatesDelimiter;
    /** @type {?} */
    DatePickerOptions.prototype.showTime;
    /** @type {?} */
    DatePickerOptions.prototype.showType;
    /** @type {?} */
    DatePickerOptions.prototype.dateFormat;
    /** @type {?|undefined} */
    DatePickerOptions.prototype.maxDate;
    /** @type {?|undefined} */
    DatePickerOptions.prototype.minDate;
    /** @type {?|undefined} */
    DatePickerOptions.prototype.showWeekNumbers;
    /** @type {?|undefined} */
    DatePickerOptions.prototype.viewType;
    /** @type {?|undefined} */
    DatePickerOptions.prototype.firstDayOfWeek;
}
/** @type {?} */
export const DatePickerDefaultOptions = {
    disabled: false,
    readonly: false,
    editable: true,
    locale: 'zh-cn',
    dateRange: false,
    dateRangeDatesDelimiter: '~',
    showTime: false,
    showType: ShowType.all,
    dateFormat: '',
    placeholder: '',
    maxDate: {
        year: 2030,
        month: 12,
        day: 31
    },
    minDate: {
        year: 1840,
        month: 1,
        day: 1
    },
    useDefault: false,
    returnType: 'String',
    returnFormat: '',
    showWeekNumbers: false,
    viewType: 'switch',
    showPresent: false,
    hourStep: 1,
    minuteStep: 1,
    secondStep: 1,
    firstDayOfWeek: 'mo'
};
/**
 * @record
 */
export function ComboListOptions() { }
if (false) {
    /** @type {?|undefined} */
    ComboListOptions.prototype.disabled;
    /** @type {?|undefined} */
    ComboListOptions.prototype.readonly;
    /** @type {?|undefined} */
    ComboListOptions.prototype.editable;
    /** @type {?|undefined} */
    ComboListOptions.prototype.panelWidth;
    /** @type {?|undefined} */
    ComboListOptions.prototype.panelHeight;
    /** @type {?|undefined} */
    ComboListOptions.prototype.data;
    /** @type {?} */
    ComboListOptions.prototype.idField;
    /** @type {?} */
    ComboListOptions.prototype.valueField;
    /** @type {?} */
    ComboListOptions.prototype.textField;
    /** @type {?|undefined} */
    ComboListOptions.prototype.uri;
    /** @type {?|undefined} */
    ComboListOptions.prototype.multiSelect;
    /** @type {?|undefined} */
    ComboListOptions.prototype.selectedValues;
    /** @type {?|undefined} */
    ComboListOptions.prototype.autoWidth;
    /** @type {?} */
    ComboListOptions.prototype.enableCancelSelected;
    /** @type {?|undefined} */
    ComboListOptions.prototype.enableClear;
    /** @type {?|undefined} */
    ComboListOptions.prototype.itemTemplate;
    /** @type {?|undefined} */
    ComboListOptions.prototype.nosearch;
    /** @type {?|undefined} */
    ComboListOptions.prototype.showDisabledItem;
    /** @type {?|undefined} */
    ComboListOptions.prototype.disabledField;
}
/** @type {?} */
export const ComboListDefaultOptions = {
    disabled: false,
    readonly: false,
    editable: true,
    placeholder: '请选择',
    panelWidth: 300,
    panelHeight: 'auto',
    autoWidth: true,
    data: [],
    idField: 'id',
    multiSelect: false,
    uri: '',
    textField: '',
    valueField: '',
    enableCancelSelected: true,
    enableClear: true,
    itemTemplate: null,
    nosearch: false,
    showDisabledItem: true,
    disabledField: 'disabled'
};
/**
 * @record
 */
export function NumberSpinnerOptions() { }
if (false) {
    /** @type {?|undefined} */
    NumberSpinnerOptions.prototype.disabled;
    /** @type {?|undefined} */
    NumberSpinnerOptions.prototype.readonly;
    /** @type {?|undefined} */
    NumberSpinnerOptions.prototype.editable;
    /** @type {?|undefined} */
    NumberSpinnerOptions.prototype.bigNumber;
    /** @type {?|undefined} */
    NumberSpinnerOptions.prototype.precision;
    /** @type {?|undefined} */
    NumberSpinnerOptions.prototype.max;
    /** @type {?|undefined} */
    NumberSpinnerOptions.prototype.min;
    /** @type {?|undefined} */
    NumberSpinnerOptions.prototype.step;
    /** @type {?|undefined} */
    NumberSpinnerOptions.prototype.canNull;
    /** @type {?|undefined} */
    NumberSpinnerOptions.prototype.groupSeparator;
    /** @type {?|undefined} */
    NumberSpinnerOptions.prototype.showZero;
    /** @type {?|undefined} */
    NumberSpinnerOptions.prototype.showButton;
}
/** @type {?} */
export const NumberSpinnerDefaultOptions = {
    disabled: false,
    readonly: false,
    editable: true,
    bigNumber: false,
    placeholder: '',
    precision: 0,
    step: 1,
    canNull: true,
    groupSeparator: ',',
    showButton: true,
    showZero: true,
    textAlign: ''
};
/** @type {?} */
export const SwitchDefaultOptions = {
    readonly: false,
    disabled: false,
    square: false,
};
/** @type {?} */
export const TimePickerDefaultOptions = {
    readonly: false,
    disabled: false,
    hourStep: 1,
    minuteStep: 1,
    secondStep: 1,
    format: 'HH:mm:ss',
    placeholder: '',
    use12Hours: false,
    showHeader: false
};
/** @type {?} */
export const ComboLookupDefaultOptions = {
    autoWidth: false,
    readonly: false,
    disabled: false,
    editable: false,
    panelHeight: 300,
    panelWidth: 400,
    enableClear: true,
    enableCancelSelected: true,
    nosearch: false,
    showHeader: true,
    treeToList: false,
    navTreeToList: false,
    separator: ','
};
/** @type {?} */
export const LanguageTextboxDefaultOptions = {
    languages: [],
    readonly: false,
    disabled: false,
    enableClear: true
};
/** @type {?} */
export const EnumEditorDefaultOptions = {
    outType: 0,
    readonly: false,
    editable: false,
    disabled: false,
    enableClear: true,
    textField: 'name',
    valueField: 'value'
};
/** @type {?} */
export const FilterEditorDefaultOptions = {
    columns: [],
    showCode: false,
    showSql: false,
    disabled: false,
    readonly: false,
    editable: false,
    enableClear: false,
    fieldEditor: { type: 'select' },
    enableExpress: false,
    returnType: 'string'
};
/** @type {?} */
export const SortEditorDefaultOptions = {
    columns: [],
    enableClear: false,
    disabled: false,
    readonly: false,
    editable: false,
    fieldEditor: { type: 'select' },
    returnType: 'string'
};
