// @gsp-svc/cloudprint/esm2015/lib/functions/cloudprint.function.register.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { PrintFunctionsOfFiltervalue } from './cloudprint.function.filtervalue';
import { PrintFunctionsOfFormat } from './cloudprint.function.format';
import { Subject } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { PrintFunctionsOfConvert } from './cloudprint.function.convert';
export class CloudprintFunctionRegister {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.httpoptions = {};
        this.httpoptions['headers'] = {
            SessionId: '',
            'Content-Type': 'application/json'
        };
    }
    /**
     * 初始化自定义函数
     * @param {?} configUrl
     * @return {?}
     */
    initCustomFunction(configUrl) {
        // 内置函数-数值过滤
        /** @type {?} */
        const funcs = new PrintFunctionsOfFiltervalue();
        // 内置函数-格式化
        /** @type {?} */
        const funcFormat = new PrintFunctionsOfFormat();
        // 内置函数-格式化
        /** @type {?} */
        const funcConvert = new PrintFunctionsOfConvert();
        // 这里默认的GSP内置函数分组为GSP
        if (Stimulsoft.Report.Dictionary.StiFunctions.getFunctions('IGixFucs').length === 0) {
            // tslint:disable-next-line:forin
            for (const f in funcs) {
                /** @type {?} */
                const func = funcs[f];
                Stimulsoft.Report.Dictionary.StiFunctions.addFunction(func.Category, func.GroupFunctionName, func.FunctionName, func.Description, func.TypeOfFunction, func.ReturnType, func.ReturnDescription, func.ArgumentTypes, func.ArgumentNames, func.ArgumentDescriptions, func.Function);
            }
            // tslint:disable-next-line:forin
            for (const f in funcFormat) {
                /** @type {?} */
                const func = funcFormat[f];
                Stimulsoft.Report.Dictionary.StiFunctions.addFunction(func.Category, func.GroupFunctionName, func.FunctionName, func.Description, func.TypeOfFunction, func.ReturnType, func.ReturnDescription, func.ArgumentTypes, func.ArgumentNames, func.ArgumentDescriptions, func.Function);
            }
            // tslint:disable-next-line:forin
            for (const f in funcConvert) {
                /** @type {?} */
                const func = funcConvert[f];
                Stimulsoft.Report.Dictionary.StiFunctions.addFunction(func.Category, func.GroupFunctionName, func.FunctionName, func.Description, func.TypeOfFunction, func.ReturnType, func.ReturnDescription, func.ArgumentTypes, func.ArgumentNames, func.ArgumentDescriptions, func.Function);
            }
        }
        // 自定义函数todo:最终自定义函数从后台取配置文件里的解析
        return this.http.get(configUrl, this.httpoptions).toPromise()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            if (data && data.length > 0) {
                return this.addJsFuncPlugins(-1, data)
                    .then((/**
                 * @return {?}
                 */
                () => {
                    if (!data) {
                        return;
                    }
                    /** @type {?} */
                    const allFuncs = [];
                    // tslint:disable-next-line:forin
                    if (data.length > 0) {
                        for (const i in data) {
                            allFuncs.push(window[((/** @type {?} */ (data[i]))).key]);
                        }
                    }
                    if (Stimulsoft.Report.Dictionary.StiFunctions.getFunctions('CustomFucs').length === 0 && allFuncs.length > 0) {
                        // tslint:disable-next-line:forin
                        for (const f in allFuncs) {
                            /** @type {?} */
                            const item = allFuncs[f];
                            // tslint:disable-next-line:forin
                            for (const f2 in item) {
                                /** @type {?} */
                                const func = item[f2];
                                Stimulsoft.Report.Dictionary.StiFunctions.addFunction('CustomFucs', func.GroupFunctionName, func.FunctionName, func.Description, func.TypeOfFunction, func.ReturnType, func.ReturnDescription, func.ArgumentTypes, func.ArgumentNames, func.ArgumentDescriptions, func.Function);
                            }
                        }
                    }
                }));
            }
        }));
    }
    /**
     * 添加js业务函数插件
     * @private
     * @param {?} i
     * @param {?} data
     * @return {?}
     */
    addJsFuncPlugins(i, data) {
        if (!data && data.length == 0) {
            /** @type {?} */
            const sub = new Subject();
            sub.complete(); // 没有这句话toPromise()里的内容不跑
            return sub.toPromise();
        }
        /** @type {?} */
        const loader = window['System'];
        if (i < data.length - 1) {
            i++;
            this.addJsFuncPlugins(i, data);
        }
        return loader.import(data[i].path);
    }
}
CloudprintFunctionRegister.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CloudprintFunctionRegister.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    CloudprintFunctionRegister.prototype.httpoptions;
    /**
     * @type {?}
     * @private
     */
    CloudprintFunctionRegister.prototype.http;
}
