// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/dimension-serializer/dimension-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/dimension-serializer/dimension-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Dimension } from '../../param-dimension/dimension';
import { DimensionItemSerializer } from './dimension-item-serializer';
var DimensionSerializer = /** @class */ (function () {
    function DimensionSerializer() {
    }
    /**
     * @param {?} dimension
     * @return {?}
     */
    DimensionSerializer.prototype.serialize = /**
     * @param {?} dimension
     * @return {?}
     */
    function (dimension) {
        /** @type {?} */
        var jobj = new Object();
        if (!dimension) {
            return null;
        }
        jobj['id'] = dimension.id;
        jobj['code'] = dimension.code;
        jobj['name'] = dimension.name;
        if (dimension.dimensionItems) {
            jobj['dimensionItems'] = [];
            /** @type {?} */
            var serializer_1 = new DimensionItemSerializer();
            dimension.dimensionItems.forEach((/**
             * @param {?} dimensionItem
             * @return {?}
             */
            function (dimensionItem) {
                jobj['dimensionItems'].push(serializer_1.serialize(dimensionItem));
            }));
        }
        return jobj;
    };
    /**
     * @param {?} jobj
     * @return {?}
     */
    DimensionSerializer.prototype.deserialize = /**
     * @param {?} jobj
     * @return {?}
     */
    function (jobj) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        var dimension = new Dimension();
        dimension.id = jobj['id'];
        dimension.code = jobj['code'];
        dimension.name = jobj['name'];
        if (jobj['dimensionItems']) {
            dimension.dimensionItems = new Array();
            /** @type {?} */
            var serializer_2 = new DimensionItemSerializer();
            jobj['dimensionItems'].forEach((/**
             * @param {?} dimensionItemJobj
             * @return {?}
             */
            function (dimensionItemJobj) {
                dimension.dimensionItems.push(serializer_2.deserialize(dimensionItemJobj));
            }));
        }
        return dimension;
    };
    return DimensionSerializer;
}());
export { DimensionSerializer };
