// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/dimension-serializer/dimension-item-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/dimension-serializer/dimension-item-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DimensionItem } from '../../param-dimension/dimension-item';
import { SerializerUtils } from '../../utils/serializer-utils';
var DimensionItemSerializer = /** @class */ (function () {
    function DimensionItemSerializer() {
    }
    /**
     * @param {?} dimensionItem
     * @return {?}
     */
    DimensionItemSerializer.prototype.serialize = /**
     * @param {?} dimensionItem
     * @return {?}
     */
    function (dimensionItem) {
        /** @type {?} */
        var jobj = new Object();
        if (!dimensionItem) {
            return null;
        }
        jobj['id'] = dimensionItem.id;
        jobj['index'] = dimensionItem.index;
        jobj['code'] = dimensionItem.code;
        jobj['name'] = dimensionItem.name;
        if (dimensionItem.dataType) {
            /** @type {?} */
            var context = SerializerUtils.getContext();
            /** @type {?} */
            var serializer = context.getDataTypeSerializerFactory().get(dimensionItem.dataType.kind);
            jobj['dataType'] = serializer.serialize(dimensionItem.dataType, context);
            //dataType作为字符串存
            jobj['dataType'] = JSON.stringify(jobj['dataType']);
        }
        return jobj;
    };
    /**
     * @param {?} jobj
     * @return {?}
     */
    DimensionItemSerializer.prototype.deserialize = /**
     * @param {?} jobj
     * @return {?}
     */
    function (jobj) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        var dimensionItem = new DimensionItem();
        dimensionItem.id = jobj['id'];
        dimensionItem.index = (/** @type {?} */ (jobj['index']));
        dimensionItem.code = jobj['code'];
        dimensionItem.name = jobj['name'];
        if (jobj['dataType']) {
            //dataType作为字符串存
            jobj['dataType'] = JSON.parse(jobj['dataType']);
            /** @type {?} */
            var context = SerializerUtils.getContext();
            /** @type {?} */
            var kind = jobj['dataType']['kind'];
            /** @type {?} */
            var serializer = context.getDataTypeSerializerFactory().get(kind);
            dimensionItem.dataType = (/** @type {?} */ (serializer.deserialize(jobj['dataType'], context)));
        }
        return dimensionItem;
    };
    return DimensionItemSerializer;
}());
export { DimensionItemSerializer };
