// @farris/mobile-devkit/esm5/lib/repository/data_change_history.js
import * as tslib_1 from "tslib";
import { DataChangeType } from '../core';
var DataChangeHistory = /** @class */ (function () {
    function DataChangeHistory() {
        this.history = [];
        this.cacheData = [];
    }
    DataChangeHistory.prototype.addChange = function (dataChange) {
        var changeType = DataChangeType[dataChange.changeType];
        this["on" + changeType + "Data"](dataChange);
    };
    DataChangeHistory.prototype.addChanges = function (dataChange) {
        var _this = this;
        dataChange.forEach(function (change) { return _this.addChange(change); });
    };
    DataChangeHistory.prototype.clear = function () {
        this.history.splice(0, this.history.length);
    };
    DataChangeHistory.prototype.clearByIds = function (ids) {
        this.history = this.history.filter(function (item) {
            var e_1, _a;
            if (item.fpath && item.fpath !== '/' && item.fpath.includes('/')) {
                try {
                    for (var ids_1 = tslib_1.__values(ids), ids_1_1 = ids_1.next(); !ids_1_1.done; ids_1_1 = ids_1.next()) {
                        var id = ids_1_1.value;
                        var include = item.fpath.split('/').includes(id);
                        return !include;
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (ids_1_1 && !ids_1_1.done && (_a = ids_1.return)) _a.call(ids_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            else {
                return !ids.includes(item.dataId);
            }
        });
    };
    DataChangeHistory.prototype.isChanged = function () {
        return this.history.length > 0;
    };
    DataChangeHistory.prototype.clearCache = function () {
        this.cacheData.splice(0, this.cacheData.length);
    };
    DataChangeHistory.prototype.getCacheData = function () {
        return this.cacheData;
    };
    DataChangeHistory.prototype.onAddData = function (dataChange) {
        this.history.push(dataChange);
    };
    DataChangeHistory.prototype.onEditData = function (dataChange) {
        this.cacheData.push(dataChange);
    };
    DataChangeHistory.prototype.onAppendData = function (dataChange) {
        this.history.push(dataChange);
        this.cacheData.push(dataChange);
    };
    DataChangeHistory.prototype.onDeleteData = function (dataChange) {
        var index = this.history.findIndex(function (item) { return item.dataId === dataChange.dataId && item.changeType === DataChangeType.Add; });
        if (index >= 0) {
            this.history.splice(index, 1);
        }
        else {
            this.history.push(dataChange);
        }
    };
    return DataChangeHistory;
}());
export { DataChangeHistory };
