// @farris/mobile-devkit/esm5/lib/repository/pagination_manager.js
import { FieldMetadataUtil } from '../entity/index';
// tslint:disable: no-bitwise
var PaginationManager = /** @class */ (function () {
    function PaginationManager(entityType, paginationConfig) {
        this.entityType = entityType;
        this.paginationConfig = paginationConfig;
        if (this.paginationConfig === null || this.paginationConfig === undefined) {
            this.paginationConfig = this.getNgListProperties();
        }
        // 兼容老表单，将之前的主表分页信息展开到分页配置根中
        this.expandMainEntityConfig();
        this.deleteMainEntityConfig();
    }
    /**
     * 主表分页信息展开到分页配置根中
     */
    PaginationManager.prototype.expandMainEntityConfig = function () {
        var entityName = this.entityType.name;
        if (this.paginationConfig.hasOwnProperty(entityName)) {
            var entityConfig = this.paginationConfig[entityName];
            this.paginationConfig = Object.assign(this.paginationConfig, entityConfig);
        }
    };
    /**
     * 删除主表实体配置信息
     */
    PaginationManager.prototype.deleteMainEntityConfig = function () {
        delete this.paginationConfig[this.entityType.name];
    };
    Object.defineProperty(PaginationManager.prototype, "pagination", {
        /**
         * 获取分页信息
         */
        get: function () {
            return this.paginationConfig;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 获取分页信息
     * @param path 路径
     * @param defaultValue 默认值
     */
    PaginationManager.prototype.getPaginationConfigByPath = function (path, defaultValue) {
        if (!path || path === '/') {
            return this.paginationConfig;
        }
        if (typeof path !== 'string') {
            throw new Error('路径必须为字符串！');
        }
        path = path.substring(1);
        var paths = path.split('/').filter(function (item) { return !!item && item.trim().length > 0; });
        var config = this.paginationConfig;
        paths.forEach(function (item) {
            if (config && config.hasOwnProperty(item)) {
                config = config[item];
            }
            else {
                config = null;
            }
        });
        return !!config ? config : typeof defaultValue !== 'undefined' ? defaultValue : undefined;
    };
    /**
     * 设置分页信息
     * @param path 路径
     * @param value 值
     */
    PaginationManager.prototype.setPaginationConfigByPath = function (path, value) {
        if (!Array.isArray(path)) {
            path = path.toString().match(/[^/[\]]+/g) || [];
        }
        path.slice(0, -1).reduce(function (prev, current, index) {
            return Object(prev[current]) === prev[current]
                ? prev[current]
                : prev[current] = Math.abs(path[index + 1]) >> 0 === +path[index + 1]
                    ? []
                    : {};
        }, this.paginationConfig)[path[path.length - 1]] = value;
        return this.paginationConfig;
    };
    /**
     * 递归获取当前实体的所有NgList属性
     * @param defaultPageSize defaultPageSize
     */
    PaginationManager.prototype.getNgListProperties = function (defaultPageSize) {
        if (defaultPageSize === void 0) { defaultPageSize = 0; }
        var getChilds = function (objectType) {
            var listProperties = FieldMetadataUtil.getNgList(objectType);
            var result = {};
            if (Object.keys(listProperties).length < 1) {
                return result;
            }
            Object.keys(listProperties).forEach(function (prop) {
                var itemTypeName = listProperties[prop].dataField;
                // 去掉尾部的s
                if (itemTypeName.endsWith('s')) {
                    itemTypeName = itemTypeName.substring(0, itemTypeName.length - 1);
                }
                result[itemTypeName] = {
                    pageSize: defaultPageSize
                };
                var child = getChilds(listProperties[prop].type);
                if (child !== null && Object.keys(child).length > 0) {
                    result = Object.assign({}, result, child);
                }
            });
            return result;
        };
        var childs = getChilds(this.entityType);
        var root = Object.assign({}, { pageSize: defaultPageSize }, childs);
        return root;
    };
    return PaginationManager;
}());
export { PaginationManager };
