// @farris/mobile-devkit/esm5/lib/repository/repository.js
/*
 * @Author: Witt
 * @Date: 2018-10-12 15:37:11
 * @Last Modified by: aalizzwell
 * @Last Modified time: 2019-09-03 19:10:44
 * @todo 待优化问题
 * 1、apiUrl是否应该在基类中，子类中的api如何传递给基类；
 */
import * as tslib_1 from "tslib";
import { DataTypeInfo } from '../entity/index';
import { createEntity, createEntities } from '../entity/index';
import { EntityCollection } from './entity_collection';
import { PaginationManager } from './pagination_manager';
import { DataChangeHistory } from './data_change_history';
var Repository = /** @class */ (function () {
    /**
     * 构造函数
     */
    function Repository() {
        /**
         * 用户分页配置信息
         */
        this.paginationInfo = null;
    }
    Repository.prototype.init = function () {
        this.entityTypeInfo = new DataTypeInfo(this.entityType);
        this.entityCollection = new EntityCollection(this.entityType);
        this.entityCollection.entityTypeInfo = this.entityTypeInfo;
        this.dataChangeHistory = new DataChangeHistory();
    };
    Object.defineProperty(Repository.prototype, "changes", {
        get: function () {
            return this.entityCollection.changes;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Repository.prototype, "primaryKey", {
        /**
         * 实体主键
         */
        get: function () {
            return this.entityCollection.primaryKey;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Repository.prototype, "entityCollectionChange", {
        /**
         * 实体变更集合
         */
        get: function () {
            return this.entityCollection.collectionChanged;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 重置状态
     */
    Repository.prototype.reset = function () {
        this.entityCollection.clear();
    };
    /**
     * 创建实体
     */
    Repository.prototype.buildEntity = function (data) {
        var entity = createEntity(this.entityType, data);
        return entity;
    };
    /**
     * 批量创建实体
     */
    Repository.prototype.buildEntities = function (listData) {
        var entities = createEntities(this.entityType, listData);
        return entities;
    };
    /**
     * 初始化分页配置
     * @param config 用户分页配置
     */
    Repository.prototype.setPaginationConfig = function (config) {
        this.paginationManager = new PaginationManager(this.entityType, config);
        var _a = (this.paginationManager.getPaginationConfigByPath('/') || {}).pageSize, pageSize = _a === void 0 ? 0 : _a;
        // tslint:disable-next-line: max-line-length
        this.entityCollection.paginationInfo = Object.assign({ pageSize: pageSize }, this.entityCollection.paginationInfo, this.paginationManager.pagination);
        // 无需再单独设置一次pageSize，减少一次变更
        // this.entityCollection.pageSize = pageSize;
    };
    /**
     * 设置分页
     */
    Repository.prototype.setPaginationInfo = function (paginationInfo) {
        this.paginationInfo = tslib_1.__assign({}, this.paginationInfo, paginationInfo);
    };
    return Repository;
}());
export { Repository };
