// @gsp-lcm/bo-rt-service/fesm2015/gsp-lcm-bo-rt-service.js
import { of } from 'rxjs/observable/of';
import { Injectable, NgModule } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/business-object-rt.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const PROJECTS_DIRECTORY = '/api/dev/main/v1.0/business-object-trees';
/** @type {?} */
const url = '/api/runtime/sys/v1.0/business-objects/';
class BusinessObjectRtService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @return {?}
     */
    getBusinessObjectTree() {
        return this.http.get(PROJECTS_DIRECTORY).pipe();
    }
    /**
     * @param {?} boId
     * @return {?}
     */
    getBusinessObject(boId) {
        /** @type {?} */
        const getUrl = url + boId;
        return this.http.get(getUrl).pipe();
    }
    /**
     * @return {?}
     */
    getBusinessObjects() {
        return this.http.get(url).pipe();
    }
    /**
     * 保存
     * @param {?} postData
     * @return {?}
     */
    save(postData) {
        return this.http.post(url, postData).pipe();
    }
    /**
     * @param {?} postData
     * @return {?}
     */
    saveChanges(postData) {
        return this.http.put(url, postData).pipe();
    }
    /**
     * 删除并重新加载
     * @param {?} dataId
     * @return {?}
     */
    remove(dataId) {
        /** @type {?} */
        const deleteUrl = url + dataId;
        return this.http.delete(deleteUrl, null).pipe();
    }
    /**
     * @param {?} boId
     * @return {?}
     */
    getDevBasicBOInfo(boId) {
        /** @type {?} */
        const devUrl = url + 'dev-basic-boinfos/';
        /** @type {?} */
        const getUrl = devUrl + boId;
        return this.http.get(getUrl).pipe();
    }
    /**
     * @param {?} parentId
     * @return {?}
     */
    getBOListByParentId(parentId) {
        /** @type {?} */
        const getUrl = url + '?parentID=' + parentId;
        return this.http.get(getUrl).pipe();
    }
    /**
     * @private
     * @template T
     * @param {?=} operation
     * @param {?=} result
     * @return {?}
     */
    handleError(operation = 'operation', result) {
        return (/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            console.error(error); // log to console instead
            return of((/** @type {?} */ (result)));
        });
    }
}
BusinessObjectRtService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
BusinessObjectRtService.ctorParameters = () => [
    { type: HttpService }
];

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bo-rt-service.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class BORtServiceModule {
}
BORtServiceModule.decorators = [
    { type: NgModule, args: [{
                declarations: [],
                imports: [
                    CommonModule
                ],
                exports: [],
                providers: [HttpService, BusinessObjectRtService]
            },] }
];

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: gsp-lcm-bo-rt-service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { BusinessObjectRtService, BORtServiceModule };
