// @edp-bif/common-api/esm2015/lib/slim-entity/impl/AbstractRtSlimBillProcessEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/slim-entity/impl/AbstractRtSlimBillProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from "../../ref";
import { AbstractSlimBillProcessEntity } from "./AbstractSlimBillProcessEntity";
/**
 * 运行时简化单据流实体类。
 * @abstract
 */
export class AbstractRtSlimBillProcessEntity extends AbstractSlimBillProcessEntity {
    constructor() {
        super(...arguments);
        this.isEnabled = false;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        this.isEnabled = false;
        this.eventRouterConditionsId = null;
        this.mappingConditionsId = null;
        this.mappingModeId = null;
        this.saveModeId = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsId", this.eventRouterConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsId", this.mappingConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeId", this.mappingModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeId", this.saveModeId);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.eventRouterConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsId");
        this.mappingConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsId");
        this.mappingModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeId");
        this.saveModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeId");
    }
}
if (false) {
    /** @type {?} */
    AbstractRtSlimBillProcessEntity.prototype.isEnabled;
    /** @type {?} */
    AbstractRtSlimBillProcessEntity.prototype.eventRouterConditionsId;
    /** @type {?} */
    AbstractRtSlimBillProcessEntity.prototype.mappingConditionsId;
    /** @type {?} */
    AbstractRtSlimBillProcessEntity.prototype.mappingModeId;
    /** @type {?} */
    AbstractRtSlimBillProcessEntity.prototype.saveModeId;
}
