// @edp-bif/common-api/esm2015/lib/slim-entity/impl/RtSlimBillCreationProcessEntityEx.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/slim-entity/impl/RtSlimBillCreationProcessEntityEx.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from "../../ref";
import { RtSlimBillCreationProcessEntity } from "./RtSlimBillCreationProcessEntity";
/**
 * 单据生单流简化实体扩展类。
 */
export class RtSlimBillCreationProcessEntityEx extends RtSlimBillCreationProcessEntity {
    /**
     * @param {?=} entity
     * @return {?}
     */
    RtSlimBillCreationProcessEntityEx(entity) {
        if (entity != null) {
            this.id = entity.id;
            this.code = entity.code;
            this.upBillVoId = entity.upBillVoId;
            this.upBillVoMetadataId = entity.upBillVoMetadataId;
            this.downBillVoId = entity.downBillVoId;
            this.downBillVoMetadataId = entity.downBillVoMetadataId;
            this.downBillVoProxyMetadataId = entity.downBillVoProxyMetadataId;
            this.downBillVoProxyId = entity.downBillVoProxyId;
            this.eventRouterConditionsId = entity.eventRouterConditionsId;
            this.mappingConditionsId = entity.mappingConditionsId;
            this.formSettingsId = entity.formSettingsId;
            this.bizEventId = entity.bizEventId;
            this.creationRuleId = entity.creationRuleId;
            this.isEnabled = entity.isEnabled;
            this.isSupportAutomatinFlow = entity.isSupportAutomatinFlow;
            this.name = entity.name;
            this.description = entity.description;
        }
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.name = null;
        this.description = null;
        this.upBillVoName = null;
        this.downBillVoName = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "description", this.description);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillVoName", this.upBillVoName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoName", this.downBillVoName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.description = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "description");
        this.upBillVoName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillVoName");
        this.downBillVoName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoName");
    }
}
if (false) {
    /** @type {?} */
    RtSlimBillCreationProcessEntityEx.prototype.name;
    /** @type {?} */
    RtSlimBillCreationProcessEntityEx.prototype.description;
    /** @type {?} */
    RtSlimBillCreationProcessEntityEx.prototype.upBillVoName;
    /** @type {?} */
    RtSlimBillCreationProcessEntityEx.prototype.downBillVoName;
}
