// @farris/ui-editor/esm2015/lib/editor.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { EditorComponent, TINYMCE_SCRIPT_SRC } from './editor.component';
import { EditorService } from './editor.service';
import { LocaleModule } from '@farris/ui-locale';
import { NotifyModule } from '@farris/ui-notify';
import { FarrisCommonModule } from '@farris/ui-common';
//  EditorModule.forRoot('/platform/common/web/assets/tinymce/tinymce.min.js'),
export class EditorModule {
    /**
     * @param {?} path
     * @return {?}
     */
    static forRoot(path) {
        return {
            ngModule: EditorModule,
            providers: [
                { provide: TINYMCE_SCRIPT_SRC, useValue: path }
            ]
        };
    }
}
EditorModule.decorators = [
    { type: NgModule, args: [{
                declarations: [EditorComponent],
                imports: [
                    CommonModule,
                    FormsModule,
                    LocaleModule.forRoot(),
                    NotifyModule.forRoot(),
                    FarrisCommonModule
                ],
                exports: [EditorComponent],
                providers: [
                    EditorService
                ]
            },] }
];
