// @edp-bif/runtime-api/esm2015/lib/bif-manual-creation/service/bif-data-selector.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bif-manual-creation/service/bif-data-selector.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ComponentFactoryResolver, Injector } from '@angular/core';
import { Subject } from 'rxjs';
import { FrameContext, ViewModel } from '@farris/devkit';
export class AbstractBifDataSelectorService {
    constructor() {
        /**
         * 实体数据Subject
         */
        this.subject4EntityDatas = new Subject();
    }
}
AbstractBifDataSelectorService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * 实体数据Subject
     * @type {?}
     */
    AbstractBifDataSelectorService.prototype.subject4EntityDatas;
    /**
     * 数据选择器弹窗引用
     * @type {?}
     */
    AbstractBifDataSelectorService.prototype.bsModalRef4BifSelector;
    /**
     * 点击弹窗“确定”按钮操作
     * @type {?}
     */
    AbstractBifDataSelectorService.prototype.clickOk4BifSelector;
    /**
     * 点击弹窗“取消”按钮操作
     * @type {?}
     */
    AbstractBifDataSelectorService.prototype.clickClose4BifSelector;
}
export class DefaultBifDataSelectorService extends AbstractBifDataSelectorService {
    /**
     * @param {?} frameContext
     * @param {?} viewModel
     * @param {?} cfr
     * @param {?} injector
     */
    constructor(frameContext, viewModel, cfr, injector) {
        super();
        this.frameContext = frameContext;
        this.viewModel = viewModel;
        this.cfr = cfr;
        this.injector = injector;
        /**
         * 实体数据Subject
         */
        this.subject4EntityDatas = new Subject();
        /**
         * 点击弹窗“确定”按钮操作
         */
        this.clickOk4BifSelector = (/**
         * @return {?}
         */
        () => {
            //发送选中的来源单据实体列表并关闭模态框
            this.viewModel.frameContext.repository.getById(this.viewModel.bindingData.getObject().id).subscribe((/**
             * @param {?} e
             * @return {?}
             */
            (e) => {
                this.subject4EntityDatas.next([e]);
                this.subject4EntityDatas.complete();
                this.bsModalRef4BifSelector.close();
            }));
        });
        /**
         * 点击弹窗“取消”按钮操作
         */
        this.clickClose4BifSelector = (/**
         * @return {?}
         */
        () => {
            this.subject4EntityDatas.complete();
            this.bsModalRef4BifSelector.close();
        });
    }
}
DefaultBifDataSelectorService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DefaultBifDataSelectorService.ctorParameters = () => [
    { type: FrameContext },
    { type: ViewModel },
    { type: ComponentFactoryResolver },
    { type: Injector }
];
if (false) {
    /**
     * 实体数据Subject
     * @type {?}
     */
    DefaultBifDataSelectorService.prototype.subject4EntityDatas;
    /**
     * 数据选择器弹窗引用
     * @type {?}
     */
    DefaultBifDataSelectorService.prototype.bsModalRef4BifSelector;
    /**
     * 点击弹窗“确定”按钮操作
     * @type {?}
     */
    DefaultBifDataSelectorService.prototype.clickOk4BifSelector;
    /**
     * 点击弹窗“取消”按钮操作
     * @type {?}
     */
    DefaultBifDataSelectorService.prototype.clickClose4BifSelector;
    /** @type {?} */
    DefaultBifDataSelectorService.prototype.frameContext;
    /** @type {?} */
    DefaultBifDataSelectorService.prototype.viewModel;
    /** @type {?} */
    DefaultBifDataSelectorService.prototype.cfr;
    /** @type {?} */
    DefaultBifDataSelectorService.prototype.injector;
}
