// @farris/mobile-command-services/esm2015/lib/business-attachment-services/business-attachment.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/business-attachment-services/business-attachment.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EMPTY, forkJoin, of } from 'rxjs';
import { tap, switchMap, map, filter, take } from 'rxjs/operators';
// import { cloneDeep } from 'lodash-es';
import { BindingPathConverter, HttpMethods, DataChangeType } from '@farris/mobile-devkit';
import { AttachmentUtil } from './business-attachment.util';
/**
 * 附件服务
 */
class BusinessAttachmentService {
    /**
     * 构造函数
     * @param {?} viewModelContext
     * @param {?} entityService
     * @param {?} removeDataService
     * @param {?} attachDataService
     */
    constructor(viewModelContext, entityService, removeDataService, attachDataService) {
        /**
         * 默认根目录
         */
        this.defaultRootDirId = 'default-root';
        this.viewModelContext = viewModelContext;
        this.attachDataService = attachDataService;
        this.entityService = entityService;
        this.removeDataService = removeDataService;
    }
    /**
     * 默认子目录
     * @private
     * @return {?}
     */
    get defaultParentDirName() {
        return this.viewModelContext.bindingData.list.currentId;
    }
    /**
     * 获取列表数据
     * @param {?} url
     * @param {?} filters
     * @param {?} sorts
     * @return {?}
     */
    loadAttachmentRows(url, filters, sorts) {
        /** @type {?} */
        const attachContext = this.getAttachContext();
        /** @type {?} */
        const befRepository = ((/** @type {?} */ (this.viewModelContext.repository)));
        /** @type {?} */
        const sessionService = befRepository.sessionService;
        /** @type {?} */
        const apiProxy = befRepository.apiProxy;
        // 组织params
        /** @type {?} */
        const params = {};
        if (typeof filters === 'string') {
            filters = JSON.parse(filters);
        }
        if (typeof sorts === 'string') {
            sorts = JSON.parse(sorts);
        }
        /** @type {?} */
        const entityFilter = {
            FilterConditions: filters || null,
            SortConditions: sorts || null,
            IsUsePagination: false,
            Pagination: {
                PageIndex: 1,
                PageSize: 0,
                PageCount: 0,
                TotalCount: 0
            }
        };
        /** @type {?} */
        const entityFilterString = JSON.stringify(entityFilter);
        params.entityFilter = entityFilterString;
        // 组织body
        /** @type {?} */
        const body = {
            dataChange: [],
            variableChange: null
        }
        // 赋值be的url
        ;
        // 赋值be的url
        befRepository.apiProxy.setAssociatedUrl(url);
        return this.waitForBeSession().pipe(switchMap((/**
         * @return {?}
         */
        () => {
            url = `${url}/extension/query`;
            /** @type {?} */
            const requestConfig = {
                headers: sessionService.extendRequestHeaders({}),
                params: params,
                body
            };
            /** @type {?} */
            const query$ = apiProxy.httpClient.request(HttpMethods.PUT, url, requestConfig).pipe(map((/**
             * @param {?} result
             * @return {?}
             */
            (result) => {
                /** @type {?} */
                const attachListData = result.returnValue.result;
                // 赋值当前所有数据
                attachContext.setData(JSON.parse(JSON.stringify(attachListData)));
                // 赋值缓存起来
                attachContext.updateData(JSON.parse(JSON.stringify(attachListData)));
                return attachListData;
            })));
            return query$;
        })));
    }
    /**
     * @private
     * @return {?}
     */
    waitForBeSession() {
        /** @type {?} */
        const befRepository = (/** @type {?} */ (this.viewModelContext.repository));
        /** @type {?} */
        const sessionService = befRepository.sessionService;
        return sessionService.getBeSessionExisted().pipe(filter((/**
         * @param {?} beSessionExisted
         * @return {?}
         */
        (beSessionExisted) => {
            return beSessionExisted === true;
        })), take(1));
    }
    /**
     * 同步变更
     * @param {?} url
     * @return {?}
     */
    syncAttachmentRowChanges(url) {
        /** @type {?} */
        const attachContext = this.getAttachContext() || {};
        /** @type {?} */
        const changeDetails = AttachmentUtil.getChangeDetails(attachContext.oldListData, attachContext.listData);
        if (!changeDetails || changeDetails.length === 0) {
            return of(false);
        }
        /** @type {?} */
        const changedRowId = changeDetails[0].ChangeInfo.DataId;
        /** @type {?} */
        const requestInfo = {
            dataChange: changeDetails,
            variableChange: null
        };
        /** @type {?} */
        const requestConfig = {
            body: requestInfo
        };
        /** @type {?} */
        const apiProxy = ((/** @type {?} */ (this.viewModelContext.repository))).apiProxy;
        url = `${url}/extension/retrieve/${changedRowId}`;
        return apiProxy.request(HttpMethods.PUT, url, requestConfig, true).pipe(map((/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            /** @type {?} */
            const context = attachContext.context;
            /** @type {?} */
            const data = context.getData();
            // goBack之后  更新缓存后的值
            context.updateData(JSON.parse(JSON.stringify(data)));
            return e;
        })));
    }
    /**
     * 上传单个文件
     * @param {?} attachmentInfoFieldPath
     * @param {?=} rootDirId
     * @param {?=} parentDirName
     * @return {?}
     */
    uploadAndUpdateRow(attachmentInfoFieldPath, rootDirId, parentDirName) {
        /** @type {?} */
        const rootId = rootDirId ? rootDirId : this.defaultRootDirId;
        /** @type {?} */
        const formId = parentDirName ? parentDirName : this.defaultParentDirName;
        if (!rootId || !formId) {
            throw new Error('rootDirId和parentDirName不能为空，请填写');
        }
        /** @type {?} */
        const fileInfos = this.getUploadFileInfosFromContext();
        /** @type {?} */
        const attachmentInfos = AttachmentUtil.convertToAttachmentInfos(fileInfos);
        if (!attachmentInfos || attachmentInfos.length === 0) {
            alert('请先上传附件');
            return EMPTY;
        }
        // 更新服务器端
        console.log('show loading ...');
        /** @type {?} */
        const firstAttachmentInfo = AttachmentUtil.getFirstAttachmentInfo(attachmentInfos);
        /** @type {?} */
        const result$ = this.attachDataService.updateRow(attachmentInfoFieldPath, firstAttachmentInfo).pipe(tap((/**
         * @return {?}
         */
        () => {
            console.log('hide loading ...');
        })));
        return result$;
    }
    /**
     * 上传多个文件
     * @param {?} url
     * @param {?} attachmentInfoFieldPath
     * @return {?}
     */
    uploadAndBatchAddRows(url, attachmentInfoFieldPath) {
        /** @type {?} */
        const commandContext = (/** @type {?} */ (this['context']));
        // 附件相关信息
        /** @type {?} */
        const rootDirId = commandContext.eventParams.rootId;
        /** @type {?} */
        const parentDirName = commandContext.eventParams.parentDirName;
        /** @type {?} */
        const rootId = rootDirId ? rootDirId : this.defaultRootDirId;
        /** @type {?} */
        const formId = parentDirName ? parentDirName : this.defaultParentDirName;
        if (!rootId || !formId) {
            throw new Error('rootDirId和parentDirName不能为空，请填写');
        }
        /** @type {?} */
        const context = commandContext.eventParams.context;
        /** @type {?} */
        const fileInfos = this.getUploadFileInfosFromContext();
        /** @type {?} */
        const attachmentInfos = AttachmentUtil.convertToAttachmentInfos(fileInfos);
        if (!attachmentInfos || attachmentInfos.length === 0) {
            alert('请先上传附件');
            return EMPTY;
        }
        // 更新服务器端
        /** @type {?} */
        const uploadUrl = `${url}/service/batchuploadattachment`;
        return this.attachDataService.updateRows(uploadUrl, attachmentInfoFieldPath, attachmentInfos).pipe(map((/**
         * @param {?} e
         * @return {?}
         */
        e => {
            // 追加数据
            context.appendData(e);
            /** @type {?} */
            const dataIds = this.getDataIdsToRemove(context, attachmentInfoFieldPath);
            dataIds.forEach((/**
             * @param {?} dataId
             * @return {?}
             */
            dataId => {
                // 添加标记
                this.viewModelContext.repository.dataChangeHistory.addChange({ fPath: '', dataId: dataId, changeType: DataChangeType.Add });
            }));
            return e;
        })));
    }
    /**
     * 批量删除附件所在的行
     * @param {?} url
     * @param {?} attachmentInfoFieldPath
     * @return {?}
     */
    removeAttachmentRows(url, attachmentInfoFieldPath) {
        /** @type {?} */
        const context = this['context'].eventParams.context;
        if (!attachmentInfoFieldPath) {
            return EMPTY;
        }
        /** @type {?} */
        const dataIds = this.getDataIdsToRemove(context, attachmentInfoFieldPath);
        // 如果是主表
        if (attachmentInfoFieldPath.split('/').length <= 2) {
            /** @type {?} */
            const removeUrl = `${url}/extension/batchdelete`;
            return this.removeDataService.removeByBusinessIds(removeUrl, dataIds).pipe(map((/**
             * @param {?} e
             * @return {?}
             */
            (e) => {
                dataIds.forEach((/**
                 * @param {?} dataId
                 * @return {?}
                 */
                dataId => {
                    context.removeDataById(dataId);
                    // 添加标记
                    this.viewModelContext.repository.dataChangeHistory.addChange({ fPath: '', dataId: dataId, changeType: DataChangeType.Delete });
                }));
            })));
        }
        else {
            // 如果是子表
            /** @type {?} */
            const removeObservables = [];
            if (dataIds.length === 0) {
                alert('请选择要删除的文件');
            }
            /** @type {?} */
            const bindingListPath = this.getBindingListPathWithAttachments(attachmentInfoFieldPath);
            dataIds.forEach((/**
             * @param {?} dataId
             * @return {?}
             */
            (dataId) => {
                /** @type {?} */
                const removeObservable = this.removeDataService.removeByBusinessPathAndId(url, bindingListPath, dataId);
                removeObservables.push(removeObservable);
            }));
            return forkJoin(removeObservables).pipe(map((/**
             * @param {?} e
             * @return {?}
             */
            (e) => {
                console.log(e);
                return e;
            })));
        }
    }
    // #region 工具方法
    /**
     * 从上下文中中获取控件传递的附件信息
     * @private
     * @return {?}
     */
    getUploadFileInfosFromContext() {
        /** @type {?} */
        const commandContext = (/** @type {?} */ (this['context']));
        /** @type {?} */
        const uploadFileInfos = (/** @type {?} */ (commandContext.eventParams.data));
        if (!uploadFileInfos) {
            return [];
        }
        return uploadFileInfos;
    }
    /**
     * 获取要删除的附件对应的数据id数组
     * @param {?} context
     * @param {?} attachmentInfoFieldPath
     * @return {?}
     */
    getDataIdsToRemove(context, attachmentInfoFieldPath) {
        /** @type {?} */
        const attachIds = this.getAttachmentIdsToRemoveFromContext();
        /** @type {?} */
        const dataIds = [];
        attachIds.forEach((/**
         * @param {?} attachId
         * @return {?}
         */
        (attachId) => {
            // 上传删除和预览删除传递过来的fileId的key可能不一致，要做兼容
            /** @type {?} */
            const dataId = this.convertAttachmentIdToDataId(context, attachId, attachmentInfoFieldPath);
            dataIds.push(dataId);
        }));
        return dataIds;
    }
    /**
     * 从命令上下文中获取要删除附件ids
     * @return {?}
     */
    getAttachmentIdsToRemoveFromContext() {
        /** @type {?} */
        const commandContext = (/** @type {?} */ (this['context']));
        return commandContext.eventParams.data.map((/**
         * @param {?} dataItem
         * @return {?}
         */
        dataItem => {
            return Object.prototype.toString.call(dataItem) === '[object Object]' ? dataItem.metadataId : dataItem;
        }));
        // return commandContext.eventParams.data as string[];
    }
    /**
     * 根据路径获取附件字段值数组
     * @private
     * @param {?} context
     * @param {?} fileId
     * @param {?} attachmentInfoFieldPath
     * @return {?}
     */
    convertAttachmentIdToDataId(context, fileId, attachmentInfoFieldPath) {
        // 解析路径
        /** @type {?} */
        const attachInfoBindingPath = BindingPathConverter.toBindingPathArray(attachmentInfoFieldPath);
        /** @type {?} */
        const attachFieldName = attachInfoBindingPath.pop();
        /** @type {?} */
        const attachListBindingPath = attachInfoBindingPath;
        // 获取附件id数组
        // const entityListData = this.entityService.getEntityListData(attachListBindingPath);
        /** @type {?} */
        const entityListData = context.data;
        /** @type {?} */
        const targetEntityData = entityListData.find((/**
         * @param {?} entityData
         * @return {?}
         */
        (entityData) => {
            if (entityData[attachFieldName]) {
                /** @type {?} */
                const attachmentId = entityData[attachFieldName]['attachmentId'];
                if (attachmentId === fileId) {
                    return true;
                }
            }
        }));
        return targetEntityData.id;
    }
    /**
     * 获取带附件的BindingList的Path
     * @private
     * @param {?} attachmentInfoFieldPath
     * @return {?}
     */
    getBindingListPathWithAttachments(attachmentInfoFieldPath) {
        /** @type {?} */
        const attachInfoBindingPath = BindingPathConverter.toBindingPathArray(attachmentInfoFieldPath);
        attachInfoBindingPath.pop();
        /** @type {?} */
        const bindingListPath = attachInfoBindingPath;
        return '/' + bindingListPath.join('/');
    }
    /**
     * 获取附件组件上下文
     * @private
     * @return {?}
     */
    getAttachContext() {
        /** @type {?} */
        const commandContext = (/** @type {?} */ (this['context']));
        /** @type {?} */
        const attachContext = commandContext && commandContext.eventParams ? commandContext.eventParams : {};
        return attachContext;
    }
}
if (false) {
    /**
     * 默认根目录
     * @type {?}
     * @private
     */
    BusinessAttachmentService.prototype.defaultRootDirId;
    /**
     * 视图模型
     * @type {?}
     * @private
     */
    BusinessAttachmentService.prototype.viewModelContext;
    /**
     * 附件数据服务
     * @type {?}
     * @private
     */
    BusinessAttachmentService.prototype.attachDataService;
    /**
     * 实体服务
     * @type {?}
     * @private
     */
    BusinessAttachmentService.prototype.entityService;
    /**
     * 实体服务
     * @type {?}
     * @private
     */
    BusinessAttachmentService.prototype.removeDataService;
}
export { BusinessAttachmentService };
