// @farris/mobile-command-services/esm2015/lib/business-attachment-services/business-attachment-data.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/business-attachment-services/business-attachment-data.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { of } from 'rxjs';
import { map } from 'rxjs/operators';
import { HttpMethods, BindingPathConverter, HttpUtil } from '@farris/mobile-devkit';
import { BefDataPathUtil } from '@farris/mobile-bef';
import { AttachmentUtil } from './business-attachment.util';
/**
 * 附件数据服务
 */
class BusinessAttachmentDataService {
    /**
     * 实体仓库
     * @return {?}
     */
    get befReposi() {
        return (/** @type {?} */ (this.viewModelContext.repository));
    }
    /**
     * 绑定数据
     * @private
     * @return {?}
     */
    get bindingData() {
        return this.viewModelContext.bindingData;
    }
    /**
     * 构造函数
     * @param {?} viewModelContext
     */
    constructor(viewModelContext) {
        this.viewModelContext = viewModelContext;
    }
    /**
     * @return {?}
     */
    getAllBeSessions() {
        /** @type {?} */
        const beSessionsJson = window.sessionStorage.getItem("BE_SESSION_ID");
        if (!beSessionsJson) {
            return {};
        }
        /** @type {?} */
        const frmSessionId = this.getUserSessionId();
        /** @type {?} */
        const beSessions = JSON.parse(beSessionsJson);
        /** @type {?} */
        const beSessionKey = `${frmSessionId}_/api/gsp/common/v1.0/testForm_mfrm`;
        return beSessions[beSessionKey];
    }
    /**
     * 获取用户SessionId
     * @return {?}
     */
    getUserSessionId() {
        /** @type {?} */
        const userSessionId = window.localStorage.getItem("sessionId");
        return userSessionId;
    }
    /**
     * 扩展BeSessionId相关头信息
     * @param {?} headers
     * @return {?}
     */
    handleRequestHeaders(headers) {
        /** @type {?} */
        const frmSessionId = this.getUserSessionId();
        /** @type {?} */
        const beSessionId = this.getAllBeSessions();
        if (frmSessionId) {
            headers = HttpUtil.appendHeader(headers, "X-CAF-Runtime-CommonVariable", frmSessionId);
        }
        if (beSessionId) {
            headers = HttpUtil.appendHeader(headers, "X-CAF-Runtime-Context", beSessionId);
            headers = HttpUtil.appendHeader(headers, "SessionId", beSessionId);
        }
        headers = HttpUtil.appendHeader(headers, 'Content-Type', 'application/json');
        return headers;
    }
    /**
     * @param {?} url
     * @return {?}
     */
    extendQueryData(url) {
        /** @type {?} */
        const entityFilter = {
            FilterConditions: [],
            SortConditions: [],
            IsUsePagination: false,
            Pagination: {
                PageIndex: 1,
                PageSize: 5,
                PageCount: 0,
                TotalCount: 0
            }
        };
        /** @type {?} */
        const entityFilterString = JSON.stringify(entityFilter);
        /** @type {?} */
        const params = {
            entityFilter: entityFilterString
        };
        // const url = `/api/gsp/common/v1.0/attachmentform_mfrm/extension/query`;
        // 拓展header
        // const headers = this.handleRequestHeaders({});
        // 拓展body
        /** @type {?} */
        const requestConfig = { params };
        return this.befReposi.apiProxy.request(HttpMethods.PUT, url, requestConfig);
    }
    /**
     * 更新附件信息
     * @param {?} attachmentInfoFieldPath
     * @param {?} attachmentInfo
     * @return {?}
     */
    updateRow(attachmentInfoFieldPath, attachmentInfo) {
        /** @type {?} */
        const apiProxy = this.befReposi.apiProxy;
        /** @type {?} */
        const updateUrl = `${apiProxy.baseUrl}/service/updateattachment`;
        /** @type {?} */
        const serverAttachInfo = this.createUpdateAttachInfo(attachmentInfoFieldPath, attachmentInfo);
        /** @type {?} */
        const body = {
            updateAttachInfo: serverAttachInfo
        };
        /** @type {?} */
        const requestConfig = {
            body: body
        };
        return apiProxy.request(HttpMethods.PUT, updateUrl, requestConfig);
    }
    /**
     * 批量创建附件行数据
     * @param {?} updateUrl
     * @param {?} attachmentInfoFieldPath
     * @param {?} attachmentInfos
     * @return {?}
     */
    updateRows(updateUrl, attachmentInfoFieldPath, attachmentInfos) {
        /** @type {?} */
        const apiProxy = this.befReposi.apiProxy;
        // const updateUrl = `${apiProxy.baseUrl}/service/batchuploadattachment`;
        /** @type {?} */
        const serverAttachInfo = this.createBatchCreateAttachInfo(attachmentInfoFieldPath, attachmentInfos);
        /** @type {?} */
        const isRootEntity = serverAttachInfo.NodeCodes.length === 0;
        /** @type {?} */
        const body = {
            batchUploadInfo: serverAttachInfo
        };
        /** @type {?} */
        const requestConfig = {
            body: body
        };
        return apiProxy.request(HttpMethods.PUT, updateUrl, requestConfig, true);
    }
    /**
     * 创建服务器端需要的更新信息
     * @private
     * @param {?} attachmentInfoFieldPath
     * @param {?} attachmentInfo
     * @return {?}
     */
    createUpdateAttachInfo(attachmentInfoFieldPath, attachmentInfo) {
        /** @type {?} */
        const attachmentId = attachmentInfo.attachmentId;
        /** @type {?} */
        const parentBindingPathArray = BindingPathConverter.toBindingPathArray(attachmentInfoFieldPath);
        parentBindingPathArray.pop();
        /** @type {?} */
        const nodeCodes = BefDataPathUtil.convertToObjectCodes(parentBindingPathArray, this.bindingData);
        /** @type {?} */
        const hiretryIds = BefDataPathUtil.convertToDataIdsForUpdate(parentBindingPathArray, this.bindingData);
        /** @type {?} */
        const serverAttachInfo = {
            NodeCodes: nodeCodes,
            HiretryIds: hiretryIds,
            AttachmentIds: [attachmentId],
            AttachmentId: attachmentId
        };
        return serverAttachInfo;
    }
    /**
     * 创建服务器端需要的批量新增附件信息
     * @private
     * @param {?} attachmentInfoFieldPath
     * @param {?} attachmentInfo
     * @return {?}
     */
    createBatchCreateAttachInfo(attachmentInfoFieldPath, attachmentInfo) {
        /** @type {?} */
        const attachmentIds = AttachmentUtil.peekAttachmentIds(attachmentInfo);
        /** @type {?} */
        const parentBindingPathArray = BindingPathConverter.toBindingPathArray(attachmentInfoFieldPath);
        parentBindingPathArray.pop();
        /** @type {?} */
        const nodeCodes = BefDataPathUtil.convertToObjectCodes(parentBindingPathArray, this.bindingData);
        /** @type {?} */
        const hiretryIds = BefDataPathUtil.convertToDataIdsForAdd(parentBindingPathArray, this.bindingData);
        /** @type {?} */
        const serverAttachInfo = {
            NodeCodes: nodeCodes,
            HiretryIds: hiretryIds,
            AttachmentIds: attachmentIds,
            AttachmentId: null
        };
        return serverAttachInfo;
    }
    /**
     * 同步服务器端最新信息到客户端
     * \@todo:
     * 1、主对象批量新增时不支持
     * @return {?}
     */
    syncAttachmentInfosToClient() {
        /** @type {?} */
        const rootDataId = this.bindingData.list.currentId;
        return this.befReposi.updateEntityById(rootDataId);
    }
    /**
     * 追击主表数据到客户端
     * @param {?} listData
     * @param {?} isRootEntity
     * @return {?}
     */
    appendAttachmentInfosToClient(listData, isRootEntity) {
        if (isRootEntity === true) {
            /** @type {?} */
            const entities = this.befReposi.buildEntities(listData);
            this.befReposi.entityCollection.addEntities(entities);
            return of(listData);
        }
        else {
            return this.syncAttachmentInfosToClient().pipe(map((/**
             * @return {?}
             */
            () => {
                return listData;
            })));
        }
    }
}
if (false) {
    /**
     * ViewModel上下文
     * @type {?}
     */
    BusinessAttachmentDataService.prototype.viewModelContext;
}
export { BusinessAttachmentDataService };
