// @farris/mobile-command-services/esm2015/lib/business-attachment-services/business-attachment.util.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/business-attachment-services/business-attachment.util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetailType } from '@farris/mobile-bef';
/**
 * 附件信息处理工具类
 */
class AttachmentUtil {
    /**
     * 转换为附件信息数组
     * @param {?} fileInfos
     * @return {?}
     */
    static convertToAttachmentInfos(fileInfos) {
        if (!fileInfos) {
            return [];
        }
        /** @type {?} */
        const attachmentInfos = fileInfos.map((/**
         * @param {?} fileInfo
         * @return {?}
         */
        (fileInfo) => {
            return this.convertToAttachmentInfo(fileInfo);
        }));
        return attachmentInfos;
    }
    /**
     * 转换为附件信息
     * @param {?} fileInfo
     * @return {?}
     */
    static convertToAttachmentInfo(fileInfo) {
        /** @type {?} */
        const attachmentInfo = {
            attachmentId: fileInfo.metadataId,
            fileName: fileInfo.fileName
        };
        return attachmentInfo;
    }
    /**
     * 获取附件列表中的第一个附件
     * @param {?} attachmentInfos
     * @return {?}
     */
    static getFirstAttachmentInfo(attachmentInfos) {
        if (!attachmentInfos || attachmentInfos.length === 0) {
            return;
        }
        /** @type {?} */
        const firstBusinessAttachmentInfo = attachmentInfos[0];
        return firstBusinessAttachmentInfo;
    }
    /**
     * 提取附件id数组
     * @param {?} attachmentInfos
     * @return {?}
     */
    static peekAttachmentIds(attachmentInfos) {
        if (!attachmentInfos) {
            attachmentInfos = [];
        }
        /** @type {?} */
        const attachmentIds = attachmentInfos.map((/**
         * @param {?} attachmentInfo
         * @return {?}
         */
        (attachmentInfo) => {
            return attachmentInfo.attachmentId;
        }));
        return attachmentIds;
    }
    /**
     * 批量获取变更
     * @param {?} oldListData
     * @param {?} newListData
     * @return {?}
     */
    static getChangeDetails(oldListData, newListData) {
        /** @type {?} */
        const changeDetails = [];
        newListData.forEach((/**
         * @param {?} newData
         * @return {?}
         */
        (newData) => {
            /** @type {?} */
            let oldData = oldListData.find((/**
             * @param {?} oldData
             * @return {?}
             */
            (oldData) => {
                return oldData['id'] === newData['id'];
            }));
            if (!oldData) {
                oldData = {};
            }
            /** @type {?} */
            const changeDetail = this.getChangeDetail(oldData, newData);
            if (changeDetail) {
                changeDetails.push(changeDetail);
            }
        }));
        return changeDetails;
    }
    /**
     * 获取变更集
     * @param {?} oldData
     * @param {?} newData
     * @return {?}
     */
    static getChangeDetail(oldData, newData) {
        /** @type {?} */
        const changeDetail = {
            ChangeType: ChangeDetailType.Modify,
            ChangeInfo: {
                DataId: newData['id']
            }
        };
        /** @type {?} */
        let isChanged = false;
        Object.keys(newData).forEach((/**
         * @param {?} propName
         * @return {?}
         */
        (propName) => {
            // 移除不必要的属性的比较
            if (propName === 'id' || propName === 'nodeCode' || propName === 'childs') {
                return;
            }
            /** @type {?} */
            const oldValue = oldData[propName];
            /** @type {?} */
            const newValue = newData[propName];
            if (oldValue === undefined && typeof newValue === 'object') {
                return;
            }
            if (JSON.stringify(oldValue) !== JSON.stringify(newValue)) {
                changeDetail.ChangeInfo[propName] = newValue;
                isChanged = true;
            }
        }));
        return isChanged ? changeDetail : null;
    }
}
export { AttachmentUtil };
