// @gsp-wf/wf-process-editor/esm2015/lib/views/bill-type-help/bill-type-help.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from "@angular/core";
import { HttpService, SessionService } from "@ecp-caf/caf-common";
import { NotifyService } from "@farris/ui-notify";
import { BusinessObjectRtService } from "@gsp-lcm/bo-rt-service";
import { TranslateService } from "@ngx-translate/core";
import { map } from "rxjs/operators";
import { DocuTypeHelperEntity } from "./entity/DocuTypeHelperEntity";
export class BillTypeHelpService {
    /**
     * @param {?} boRtService
     * @param {?} httpClient
     * @param {?} httpService
     * @param {?} sessionService
     * @param {?} notify
     * @param {?} translate
     */
    constructor(boRtService, httpClient, httpService, sessionService, notify, translate) {
        this.boRtService = boRtService;
        this.httpClient = httpClient;
        this.httpService = httpService;
        this.sessionService = sessionService;
        this.notify = notify;
        this.translate = translate;
        this.documentTypeUrl = "/api/runtime/aif/v1.0/documenttype";
    }
    /**
     * @return {?}
     */
    findDocumentTypeEntites() {
        /** @type {?} */
        const body = {};
        return this.httpService.post(`${this.documentTypeUrl}/docutypehelper`, body).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        (res) => {
            return res.map((/**
             * @param {?} it
             * @return {?}
             */
            it => {
                /** @type {?} */
                const e = new DocuTypeHelperEntity();
                e.LoadFromJsonObject(it);
                return e;
            }));
        })));
    }
}
BillTypeHelpService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
BillTypeHelpService.ctorParameters = () => [
    { type: BusinessObjectRtService },
    { type: HttpService },
    { type: HttpService },
    { type: SessionService },
    { type: NotifyService },
    { type: TranslateService }
];
if (false) {
    /** @type {?} */
    BillTypeHelpService.prototype.documentTypeUrl;
    /** @type {?} */
    BillTypeHelpService.prototype.boRtService;
    /** @type {?} */
    BillTypeHelpService.prototype.httpClient;
    /** @type {?} */
    BillTypeHelpService.prototype.httpService;
    /** @type {?} */
    BillTypeHelpService.prototype.sessionService;
    /** @type {?} */
    BillTypeHelpService.prototype.notify;
    /** @type {?} */
    BillTypeHelpService.prototype.translate;
}
