// @farris/mobile-command-services/esm2015/lib/exception-services/exception.servcie.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/exception-services/exception.servcie.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpUtil, HttpStatusType } from './http';
/**
 * 实体操作服务
 */
class ExceptionService {
    /**
     * 构造函数
     * @param {?} notifyService
     * @param {?} jsBridgeService
     */
    constructor(notifyService, jsBridgeService) {
        this.notifyService = notifyService;
        this.jsBridgeService = jsBridgeService;
    }
    /**
     * 显示异常信息
     * @param {?} error
     * @return {?}
     */
    show(error) {
        this.exception(error);
    }
    /**
     * 异常处理路由方法
     * @private
     * @param {?} error
     * @return {?}
     */
    exception(error) {
        // 针对session失效做特殊处理
        if (error.message && error.message.indexOf('401') !== -1) {
            this.notifyService.error('登录信息已失效请重新打开');
            setTimeout((/**
             * @return {?}
             */
            () => { this.jsBridgeService.closeWindow(); }), 2000);
            return;
        }
        if (error && error.response) {
            this.httpErrorHandler(error.response);
        }
    }
    /**
     * http请求异常处理器
     * @private
     * @param {?} httpError
     * @return {?}
     */
    httpErrorHandler(httpError) {
        if (!httpError) {
            return;
        }
        /** @type {?} */
        const statusType = HttpUtil.getHttpStatusType(httpError.status);
        switch (statusType) {
            case HttpStatusType.ClientError:
                this.httpErrorInClient(httpError);
                break;
            case HttpStatusType.ServerError:
                this.httpErrorInServer(httpError);
                break;
            default:
                throw new Error('Get invalid status code when using httpErrorHandler method.');
        }
    }
    /**
     * 服务器端异常提示弹框
     * @private
     * @param {?} httpError
     * @return {?}
     */
    httpErrorInServer(httpError) {
        /** @type {?} */
        let error = httpError.data;
        if (typeof (error) === 'string') {
            try {
                error = JSON.parse(error);
            }
            catch (e) { }
        }
        if (!error || error.Level == null || error.Level == undefined) { // 未封装error或未设置异常等级
            this.notifyService.error(httpError.message);
            return;
        }
        /** @type {?} */
        const getNotifyServiceMethodName = this.getNotifyServiceMethodName(error.Level);
        this.notifyService[`${getNotifyServiceMethodName}`](error.Message);
    }
    /**
     * 客户端异常提示弹框
     * @private
     * @param {?} httpError
     * @return {?}
     */
    httpErrorInClient(httpError) {
        if (!httpError) {
            return;
        }
        /** @type {?} */
        const url = httpError.config && httpError.config.url && httpError.config.url;
        /** @type {?} */
        const status = httpError.status && httpError.status;
        this.notifyService.error(`${url} ${status}`);
    }
    /**
     * 获取异常处理策略
     * @param {?} exceptionLevel 异常等级
     * @return {?}
     */
    getNotifyServiceMethodName(exceptionLevel) {
        /** @type {?} */
        let exception;
        switch (exceptionLevel) {
            case 0 /* Info */:
                exception = 'info';
                break;
            case 1 /* Warning */:
                exception = 'warning';
                break;
            case 2 /* Error */:
                exception = 'error';
                break;
            case 3 /* Fatal */:
                exception = 'fatal';
                break;
            default:
                // 默认异常等级是错误
                exception = 'error';
                break;
        }
        return exception;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    ExceptionService.prototype.notifyService;
    /**
     * @type {?}
     * @private
     */
    ExceptionService.prototype.jsBridgeService;
}
export { ExceptionService };
