// @gsp-lcm/metadatart-selector4biztype/esm2015/metadata/metadatart-select4biztype/metadatainfo.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadatart-select4biztype/metadatainfo.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { GSPMetadataRTService4BizType } from '../metadata.rtservice4biztype';
export class MetadataInfoService {
    /**
     * @param {?} http
     * @param {?} gspMetadataService
     */
    constructor(http, gspMetadataService) {
        this.http = http;
        this.gspMetadataService = gspMetadataService;
        // tslint:disable-next-line:variable-name
        this.BusinessObject_Base_Path = '/api/dev/main/v1.0/business-objects';
        this.metadatas = [];
    }
    /**
     * @param {?} metadata
     * @return {?}
     */
    addMetadataRT(metadata) {
        this.metadatas.push(metadata);
    }
    /**
     * @return {?}
     */
    clearMetadataRT() {
        this.metadatas = [];
    }
    // 获取包源
    /**
     * @param {?} obId
     * @return {?}
     */
    getBusinessObject(obId) {
        /** @type {?} */
        const url = this.BusinessObject_Base_Path + '/' + obId;
        return this.http.get(url);
    }
    /**
     * @return {?}
     */
    GetMetadataTypeList() {
        return this.gspMetadataService.GetMetadataTypeList();
    }
    /**
     * @param {?} typeList
     * @return {?}
     */
    GetMetadataRTList(typeList) {
        return this.gspMetadataService.GetAllOriginMdRTListWithType(typeList);
    }
    // 获根据Id取指定元数据
    /**
     * @param {?} metadataID
     * @return {?}
     */
    GetMetadataRT(metadataID) {
        return this.gspMetadataService.GetMetadataRT(metadataID);
    }
}
MetadataInfoService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
MetadataInfoService.ctorParameters = () => [
    { type: HttpClient },
    { type: GSPMetadataRTService4BizType }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    MetadataInfoService.prototype.BusinessObject_Base_Path;
    /** @type {?} */
    MetadataInfoService.prototype.metadatas;
    /**
     * @type {?}
     * @private
     */
    MetadataInfoService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    MetadataInfoService.prototype.gspMetadataService;
}
