// @farris/ui-modal/esm2015/lib/modal-button.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Injector, ElementRef, Input } from '@angular/core';
export class ModalButtonDirective {
    /**
     * @param {?} injector
     * @param {?} el
     */
    constructor(injector, el) {
        this.injector = injector;
        this.el = el;
        this.visable = true;
    }
    /**
     * @param {?} disable
     * @return {?}
     */
    setDisabled(disable) {
        this.el.nativeElement.disabled = disable;
    }
    /**
     * @return {?}
     */
    show() {
        this.visable = true;
    }
    /**
     * @return {?}
     */
    hide() {
        this.visable = false;
    }
}
ModalButtonDirective.decorators = [
    { type: Directive, args: [{ selector: '[modal-button]' },] }
];
/** @nocollapse */
ModalButtonDirective.ctorParameters = () => [
    { type: Injector },
    { type: ElementRef }
];
ModalButtonDirective.propDecorators = {
    visable: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    ModalButtonDirective.prototype.visable;
    /**
     * @type {?}
     * @private
     */
    ModalButtonDirective.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    ModalButtonDirective.prototype.el;
}
