// @farris/ui-modal/esm2015/lib/modal.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PerfectScrollbarModule } from '@farris/ui-perfect-scrollbar';
import { AngularDraggableModule } from '@farris/ui-draggable';
import { ModalBackdropComponent } from './modal-backdrop.component';
import { ModalDirective } from './modal.directive';
import { ModalContainerComponent } from './modal-container.component';
import { BsModalService } from './bs-modal.service';
import { PositioningService } from '@farris/ui-modal/positioning';
import { ComponentLoaderFactory } from '@farris/ui-modal/component-loader';
import { ModalButtonDirective } from './modal-button.directive';
import { FAreaResponseModule } from '@farris/ui-area-response';
import { FarrisTooltipModule } from '@farris/ui-tooltip';
export class ModalModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: ModalModule,
            providers: [
                ComponentLoaderFactory,
                BsModalService,
                PositioningService
            ]
        };
    }
}
ModalModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    PerfectScrollbarModule,
                    AngularDraggableModule,
                    FAreaResponseModule,
                    FarrisTooltipModule
                ],
                declarations: [
                    ModalDirective,
                    ModalBackdropComponent,
                    ModalContainerComponent,
                    ModalButtonDirective
                ],
                exports: [
                    ModalBackdropComponent, ModalDirective
                ],
                entryComponents: [ModalBackdropComponent, ModalContainerComponent]
            },] }
];
