// @gsp-sys/rac-common/esm2015/lib/common/service/authobj.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { RestfulService } from '@farris/devkit';
import { RacCommonUtil } from '../util/rac-util';
import { of } from 'rxjs/observable/of';
import { MessagerService } from '@farris/ui-messager';
/**
 * 取数服务
 */
export class AuthObjService {
    /**
     * @param {?} rest
     * @param {?} util
     * @param {?} msgService
     */
    constructor(rest, util, msgService) {
        this.rest = rest;
        this.util = util;
        this.msgService = msgService;
    }
    /**
     * 获取指定权限对象、数据操作下关联的权限字段
     * @param {?} authObjId
     * @param {?} authOpId
     * @return {?}
     */
    getAuthEntryByOp(authObjId, authOpId) {
        /** @type {?} */
        const url = this.util.getAuthorizationPath() + `/` + authObjId + `/` + authOpId;
        /** @type {?} */
        const headers = this.util.getHeader();
        return this.rest.get(url, null, headers);
    }
    /**
     * @return {?}
     */
    getAll() {
        /** @type {?} */
        const headers = this.util.getHeader();
        /** @type {?} */
        const url = this.util.getAuthorizationPath() + '/designtime';
        return this.rest.get(url, null, headers);
    }
    /**
     * @return {?}
     */
    getAllReAssign() {
        /** @type {?} */
        const headers = this.util.getHeader();
        /** @type {?} */
        const url = this.util.getAuthorizationPath() + '/reassign';
        return this.rest.get(url, null, headers);
    }
    /**
     * 异常处理
     * @param {?} error
     * @return {?}
     */
    handlerError(error) {
        console.log(error);
        this.msgService.error(error.error.Message);
    }
    /**
     * @private
     * @template T
     * @param {?=} operation
     * @param {?=} result
     * @return {?}
     */
    handleError(operation = 'operation', result) {
        return (/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            console.error(error); // log to console instead
            return of((/** @type {?} */ (result)));
        });
    }
}
AuthObjService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AuthObjService.ctorParameters = () => [
    { type: RestfulService },
    { type: RacCommonUtil },
    { type: MessagerService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthObjService.prototype.rest;
    /**
     * @type {?}
     * @private
     */
    AuthObjService.prototype.util;
    /**
     * @type {?}
     * @private
     */
    AuthObjService.prototype.msgService;
}
