// @gsp-sys/rac-common/esm2015/lib/common/service/authfield-service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { RestfulService } from '@farris/devkit';
import { RacCommonUtil } from '../util/rac-util';
import { of } from 'rxjs/observable/of';
import { MessagerService } from '@farris/ui-messager';
/**
 * 取数服务
 */
export class AuthfieldService {
    /**
     * @param {?} rest
     * @param {?} util
     * @param {?} msgService
     */
    constructor(rest, util, msgService) {
        this.rest = rest;
        this.util = util;
        this.msgService = msgService;
    }
    /**
     * 根据权限字段的标识获取权限字段
     * @param {?} authfield 权限字段标识
     * @return {?}
     */
    getbyId(authfield) {
        /** @type {?} */
        const headers = this.util.getHeader();
        /** @type {?} */
        const url = this.util.getAuthfieldPath() + `/` + authfield;
        return this.rest.get(url, null, headers);
    }
    /**
     * 异常处理
     * @param {?} error
     * @return {?}
     */
    handlerError(error) {
        console.log(error);
        this.msgService.error(error.error.Message);
    }
    /**
     * @private
     * @template T
     * @param {?=} operation
     * @param {?=} result
     * @return {?}
     */
    handleError(operation = 'operation', result) {
        return (/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            console.error(error); // log to console instead
            return of((/** @type {?} */ (result)));
        });
    }
}
AuthfieldService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AuthfieldService.ctorParameters = () => [
    { type: RestfulService },
    { type: RacCommonUtil },
    { type: MessagerService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthfieldService.prototype.rest;
    /**
     * @type {?}
     * @private
     */
    AuthfieldService.prototype.util;
    /**
     * @type {?}
     * @private
     */
    AuthfieldService.prototype.msgService;
}
