// @edp-bif/common-api/esm2015/lib/entity/impl/BillTrackingSettings.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/BillTrackingSettings.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '../../ref';
import { BillTrackingDataShownMode } from '../BillTrackingDataShownMode';
import { BillTrackingNodeShownMode } from '../BillTrackingNodeShownMode';
export class BillTrackingSettings extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.billTrackingNodeShownMode = BillTrackingNodeShownMode.DEFAULT;
        this.billTrackingDataShownMode = BillTrackingDataShownMode.DEFAULT;
    }
    ;
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.billTrackingNodeShownMode = BillTrackingNodeShownMode.DEFAULT;
        this.billNodeName = null;
        this.billNodeShownStyle = null;
        this.summaryFieldsInfo = null;
        this.billTrackingDataShownMode = BillTrackingDataShownMode.DEFAULT;
        this.customFieldContent = null;
        this.customFormContent = null;
        this.version = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "billNodeName", this.billNodeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billTrackingNodeShownMode", this.billTrackingNodeShownMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billNodeShownStyle", this.billNodeShownStyle);
        BpmnModelJsonHelper.SetAttributeValue(obj, "summaryFieldsInfo", this.summaryFieldsInfo);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billTrackingDataShownMode", this.billTrackingDataShownMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "customFieldContent", this.customFieldContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "customFormContent", this.customFormContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "version", this.version);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.billNodeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billNodeName");
        this.billTrackingNodeShownMode = BillTrackingDataShownMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billTrackingNodeShownMode")];
        this.billNodeShownStyle = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billNodeShownStyle");
        this.summaryFieldsInfo = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "summaryFieldsInfo");
        this.billTrackingDataShownMode = BillTrackingDataShownMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billTrackingDataShownMode")];
        this.customFieldContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "customFieldContent");
        this.customFormContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "customFormContent");
        this.version = (/** @type {?} */ (BpmnModelJsonHelper.GetAttributeValueAsString(obj, "version")));
    }
}
if (false) {
    /** @type {?} */
    BillTrackingSettings.prototype.billNodeName;
    /** @type {?} */
    BillTrackingSettings.prototype.billTrackingNodeShownMode;
    /** @type {?} */
    BillTrackingSettings.prototype.billNodeShownStyle;
    /** @type {?} */
    BillTrackingSettings.prototype.summaryFieldsInfo;
    /** @type {?} */
    BillTrackingSettings.prototype.billTrackingDataShownMode;
    /** @type {?} */
    BillTrackingSettings.prototype.customFieldContent;
    /** @type {?} */
    BillTrackingSettings.prototype.customFormContent;
    /**
     * 旧版本此字段为null，新版本为1.0
     * @type {?}
     */
    BillTrackingSettings.prototype.version;
    /* Skipping unhandled member: ;*/
}
