// @edp-bif/common-api/esm2015/lib/entity/impl/BillTrackingConfigEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/BillTrackingConfigEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Entity, BpmnModelJsonHelper } from '../../ref';
export class BillTrackingConfigEntity extends Entity {
    constructor() {
        super(...arguments);
        this.isEnabled = false;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.suCode = null;
        this.classTypeName = null;
        this.isEnabled = false;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "classTypeName", this.classTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
        this.classTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "classTypeName");
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
    }
}
if (false) {
    /** @type {?} */
    BillTrackingConfigEntity.prototype.suCode;
    /** @type {?} */
    BillTrackingConfigEntity.prototype.classTypeName;
    /** @type {?} */
    BillTrackingConfigEntity.prototype.isEnabled;
}
