// @edp-bif/common-api/esm2015/lib/entity/impl/ComponentConditionContent.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ComponentConditionContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from "../../ref";
import { ConditionContent } from './ConditionContent';
export class ComponentConditionContent extends ConditionContent {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.componentMetadataId = null;
        this.componentId = null;
        this.name = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentMetadataId", this.componentMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentId", this.componentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.componentMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentMetadataId");
        this.componentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentId");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
    }
}
if (false) {
    /** @type {?} */
    ComponentConditionContent.prototype.componentMetadataId;
    /** @type {?} */
    ComponentConditionContent.prototype.componentId;
    /** @type {?} */
    ComponentConditionContent.prototype.name;
}
