// @edp-bif/common-api/esm2015/lib/entity/impl/ConditionEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ConditionEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '../../ref';
import { ConditionType } from '../ConditionType';
/**
 * 条件实体类。
 */
export class ConditionEntity extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.isEnabled = false;
        this.remarks = null;
        this.supportAutoProcess = true;
        this.supportPushProcess = false;
        this.supportPullProcess = false;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.id = null;
        this.isEnabled = false;
        this.name = null;
        this.type = null;
        this.content = null;
        this.remarks = null;
        this.supportAutoProcess = true;
        this.supportPushProcess = false;
        this.supportPullProcess = false;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "id", this.id);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "type", this.type);
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        BpmnModelJsonHelper.SetAttributeValue(obj, "remarks", this.remarks);
        BpmnModelJsonHelper.SetAttributeValue(obj, "supportAutoProcess", this.supportAutoProcess);
        BpmnModelJsonHelper.SetAttributeValue(obj, "supportPushProcess", this.supportPushProcess);
        BpmnModelJsonHelper.SetAttributeValue(obj, "supportPullProcess", this.supportPullProcess);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.id = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "id");
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.type = ConditionType[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "type")];
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
        this.remarks = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "remarks");
        this.supportAutoProcess = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "supportAutoProcess");
        this.supportPushProcess = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "supportPushProcess");
        this.supportPullProcess = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "supportPullProcess");
    }
}
if (false) {
    /** @type {?} */
    ConditionEntity.prototype.id;
    /** @type {?} */
    ConditionEntity.prototype.isEnabled;
    /** @type {?} */
    ConditionEntity.prototype.name;
    /** @type {?} */
    ConditionEntity.prototype.type;
    /** @type {?} */
    ConditionEntity.prototype.content;
    /** @type {?} */
    ConditionEntity.prototype.remarks;
    /** @type {?} */
    ConditionEntity.prototype.supportAutoProcess;
    /** @type {?} */
    ConditionEntity.prototype.supportPushProcess;
    /** @type {?} */
    ConditionEntity.prototype.supportPullProcess;
}
