// @edp-bif/common-api/esm2015/lib/entity/impl/MappingModeEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/MappingModeEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from "../../ref";
import { MappingMode } from '../MappingMode';
/**
 * 映射方式实体类。
 */
export class MappingModeEntity extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.mappingMode = null;
        this.content = null;
        this.mappingRuleContent = null;
        this.componentContent = null;
        this.voActionContent = null;
        this.expressionContent = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingMode", this.mappingMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleContent", this.mappingRuleContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentContent", this.componentContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voActionContent", this.voActionContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "expressionContent", this.expressionContent);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.mappingMode = MappingMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingMode")];
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
        this.mappingRuleContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleContent");
        this.componentContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentContent");
        this.voActionContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voActionContent");
        this.expressionContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "expressionContent");
    }
}
if (false) {
    /** @type {?} */
    MappingModeEntity.prototype.mappingMode;
    /** @type {?} */
    MappingModeEntity.prototype.content;
    /** @type {?} */
    MappingModeEntity.prototype.mappingRuleContent;
    /** @type {?} */
    MappingModeEntity.prototype.componentContent;
    /** @type {?} */
    MappingModeEntity.prototype.voActionContent;
    /** @type {?} */
    MappingModeEntity.prototype.expressionContent;
}
