// @edp-bif/common-api/esm2015/lib/entity/impl/MappingRuleContent.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/MappingRuleContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, JsonConvertableImpl } from '../../ref';
export class MappingRuleContent extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.mappingRuleMetadataId = null;
        this.mappingRuleId = null;
        this.mappingContent = null;
        this.mappingConfig = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleMetadataId", this.mappingRuleMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleId", this.mappingRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingContent", this.mappingContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConfig", this.mappingConfig);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.mappingRuleMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleMetadataId");
        this.mappingRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleId");
        this.mappingContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingContent");
        this.mappingConfig = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConfig");
    }
}
if (false) {
    /** @type {?} */
    MappingRuleContent.prototype.mappingRuleMetadataId;
    /** @type {?} */
    MappingRuleContent.prototype.mappingRuleId;
    /** @type {?} */
    MappingRuleContent.prototype.mappingContent;
    /** @type {?} */
    MappingRuleContent.prototype.mappingConfig;
}
