// @edp-bif/common-api/esm2015/lib/entity/impl/RtBillWritebackProcessEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/RtBillWritebackProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from '../../ref';
import { AbstractBillWritebackProcessEntity } from './AbstractBillWritebackProcessEntity';
export class RtBillWritebackProcessEntity extends AbstractBillWritebackProcessEntity {
    constructor() {
        super(...arguments);
        this.isEnabled = false;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.isEnabled = false;
        this.eventRouterConditionsId = null;
        this.mappingConditionsId = null;
        this.mappingModeId = null;
        this.saveModeId = null;
        this.dtBillWritebackProcessId = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsId", this.eventRouterConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsId", this.mappingConditionsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeId", this.mappingModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeId", this.saveModeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dtBillWritebackProcessId", this.dtBillWritebackProcessId);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.eventRouterConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "eventRouterConditionsId");
        this.mappingConditionsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingConditionsId");
        this.mappingModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingModeId");
        this.saveModeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveModeId");
        this.dtBillWritebackProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dtBillWritebackProcessId");
    }
}
if (false) {
    /** @type {?} */
    RtBillWritebackProcessEntity.prototype.isEnabled;
    /** @type {?} */
    RtBillWritebackProcessEntity.prototype.eventRouterConditionsId;
    /** @type {?} */
    RtBillWritebackProcessEntity.prototype.mappingConditionsId;
    /** @type {?} */
    RtBillWritebackProcessEntity.prototype.mappingModeId;
    /** @type {?} */
    RtBillWritebackProcessEntity.prototype.saveModeId;
    /** @type {?} */
    RtBillWritebackProcessEntity.prototype.dtBillWritebackProcessId;
}
