// @edp-bif/common-api/esm2015/lib/entity/impl/SaveModeEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/SaveModeEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '../../ref';
import { SaveMode } from '../SaveMode';
/**
 * 保存模式实体类。
 */
export class SaveModeEntity extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.saveMode = null;
        this.content = null;
        this.componentContent = null;
        this.voActionContent = null;
        this.taskCenterContent = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveMode", this.saveMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        BpmnModelJsonHelper.SetAttributeValue(obj, "componentContent", this.componentContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voActionContent", this.voActionContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "taskCenterContent", this.taskCenterContent);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.saveMode = SaveMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "saveMode")];
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
        this.componentContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "componentContent");
        this.voActionContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voActionContent");
        this.taskCenterContent = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "taskCenterContent");
    }
}
if (false) {
    /** @type {?} */
    SaveModeEntity.prototype.saveMode;
    /** @type {?} */
    SaveModeEntity.prototype.content;
    /** @type {?} */
    SaveModeEntity.prototype.componentContent;
    /** @type {?} */
    SaveModeEntity.prototype.voActionContent;
    /** @type {?} */
    SaveModeEntity.prototype.taskCenterContent;
}
