// @edp-bif/common-api/esm2015/lib/entity/impl/TableSchema.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/TableSchema.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from "@edp-aif/common-api";
export class TableSchema extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.maxDisplayNumber = 2000;
        this.isMainTable = false;
        this.isDisplay = true;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.code = null;
        this.maxDisplayNumber = 2000;
        this.isMainTable = false;
        this.isDisplay = true;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "maxDisplayNumber", this.maxDisplayNumber);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isMainTable", this.isMainTable);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isDisplay", this.isDisplay);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.maxDisplayNumber = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "maxDisplayNumber");
        this.isMainTable = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isMainTable");
        this.isDisplay = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isDisplay");
    }
}
if (false) {
    /** @type {?} */
    TableSchema.prototype.code;
    /** @type {?} */
    TableSchema.prototype.maxDisplayNumber;
    /** @type {?} */
    TableSchema.prototype.isMainTable;
    /** @type {?} */
    TableSchema.prototype.isDisplay;
}
