// @edp-bif/common-api/esm2015/lib/entity/impl/ThirdSchemaInfoEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ThirdSchemaInfoEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper, Entity } from "@edp-aif/common-api";
export class ThirdSchemaInfoEntity extends Entity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.schemaCategoryId = null;
        this.schemaCategoryName = null;
        this.schemaConvertorClassTypeName = null;
        this.trackingConfigClassTypeName = null;
        this.suCode = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaCategoryId", this.schemaCategoryId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaCategoryName", this.schemaCategoryName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "schemaConvertorClassTypeName", this.schemaConvertorClassTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "trackingConfigClassTypeName", this.trackingConfigClassTypeName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "suCode", this.suCode);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.schemaCategoryId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaCategoryId");
        this.schemaCategoryName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaCategoryName");
        this.schemaConvertorClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "schemaConvertorClassTypeName");
        this.trackingConfigClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "trackingConfigClassTypeName");
        this.suCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "suCode");
    }
}
if (false) {
    /** @type {?} */
    ThirdSchemaInfoEntity.prototype.schemaCategoryId;
    /** @type {?} */
    ThirdSchemaInfoEntity.prototype.schemaCategoryName;
    /** @type {?} */
    ThirdSchemaInfoEntity.prototype.schemaConvertorClassTypeName;
    /** @type {?} */
    ThirdSchemaInfoEntity.prototype.trackingConfigClassTypeName;
    /** @type {?} */
    ThirdSchemaInfoEntity.prototype.suCode;
}
