// @farris/kendo-binding/esm2015/lib/directives/farris-filter/farris_filter_binding.directive.js
import { Directive, Injector, Input, Optional } from '@angular/core';
import { FrameContext } from '@farris/devkit';
import { FilterComponent } from '@farris/ui-filter';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
export class FarrisFilterBindingDirective {
    constructor(injector, frameContext, filterComponent) {
        this.injector = injector;
        this.frameContext = frameContext;
        this.filterComponent = filterComponent;
        /**
         * 变量编号，默认为filterConditionList
         */
        this.variableCode = 'filterConditionList';
        this.$destroy = new Subject();
    }
    ngOnDestroy() {
        if (this.$destroy) {
            this.$destroy.next();
            this.$destroy.complete();
        }
    }
    ngOnInit() {
        if (this.filterComponent) {
            this.filterComponent.conditionsChange.pipe(takeUntil(this.$destroy)).subscribe((event) => {
                const { data = '[]' } = event || {};
                this.setConditions(data);
            });
        }
    }
    setConditions(conditions) {
        this.frameContext.uiState.setPropertyValue(this.variableCode, conditions);
    }
}
FarrisFilterBindingDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-filter-binding]'
            },] }
];
/** @nocollapse */
FarrisFilterBindingDirective.ctorParameters = () => [
    { type: Injector },
    { type: FrameContext },
    { type: FilterComponent, decorators: [{ type: Optional }] }
];
FarrisFilterBindingDirective.propDecorators = {
    variableCode: [{ type: Input, args: ['variableCode',] }]
};
