// @farris/ide-devkit/esm5/lib/components/dropdown/dropdown-button.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, HostBinding, HostListener, Input } from '@angular/core';
import { BsDropdownState } from './dropdown.state';
var DropdownButtonDirective = /** @class */ (function () {
    function DropdownButtonDirective(_state, _element) {
        var _this = this;
        this._state = _state;
        this._element = _element;
        this._subscriptions = [];
        this.isDisabled = null;
        // 同步菜单的打开状态
        this._subscriptions.push(this._state.isOpenChange
            .subscribe(function (value) { return _this.isOpen = value; }));
        // 同步菜单的禁用状态
        this._subscriptions.push(this._state.isDisabledChange
            .subscribe(function (value) { return _this.isDisabled = value || null; }));
    }
    /**
     * @param {?} event
     * @return {?}
     */
    DropdownButtonDirective.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.isDisabled) {
            return;
        }
        this._state.onClick.emit(event);
    };
    /**
     * @return {?}
     */
    DropdownButtonDirective.prototype.onHover = /**
     * @return {?}
     */
    function () {
        if (this.isDisabled) {
            return;
        }
        this._state.onHover.emit();
    };
    /**
     * @return {?}
     */
    DropdownButtonDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        try {
            for (var _b = tslib_1.__values(this._subscriptions), _c = _b.next(); !_c.done; _c = _b.next()) {
                var sub = _c.value;
                sub.unsubscribe();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    DropdownButtonDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[dropdownButton]'
                },] }
    ];
    /** @nocollapse */
    DropdownButtonDirective.ctorParameters = function () { return [
        { type: BsDropdownState },
        { type: ElementRef }
    ]; };
    DropdownButtonDirective.propDecorators = {
        isDisabled: [{ type: HostBinding, args: ['class.disabled',] }, { type: Input }],
        isOpen: [{ type: HostBinding, args: ['attr.aria-expanded',] }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }],
        onHover: [{ type: HostListener, args: ['mouseover',] }]
    };
    return DropdownButtonDirective;
}());
export { DropdownButtonDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DropdownButtonDirective.prototype._subscriptions;
    /** @type {?} */
    DropdownButtonDirective.prototype.isDisabled;
    /** @type {?} */
    DropdownButtonDirective.prototype.isOpen;
    /**
     * @type {?}
     * @private
     */
    DropdownButtonDirective.prototype._state;
    /**
     * @type {?}
     * @private
     */
    DropdownButtonDirective.prototype._element;
}
