// @farris/ide-devkit/esm5/lib/components/dropdown/dropdown-item.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, HostBinding, HostListener, Input } from '@angular/core';
import { NavbarState } from '../navbar/navbar.state';
var DropdownItemDirective = /** @class */ (function () {
    function DropdownItemDirective(_elementRef, barState) {
        this._elementRef = _elementRef;
        this.barState = barState;
        this.isDisabled = null;
        this.onHover = new EventEmitter();
        this.onClick = new EventEmitter();
        this.barState.addMenuItem(this);
    }
    Object.defineProperty(DropdownItemDirective.prototype, "elementRef", {
        get: /**
         * @return {?}
         */
        function () {
            return this._elementRef;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DropdownItemDirective.prototype.onMouseOver = /**
     * @return {?}
     */
    function () {
        this.onHover.emit();
    };
    /**
     * @return {?}
     */
    DropdownItemDirective.prototype.onItemClick = /**
     * @return {?}
     */
    function () {
        if (this.isDisabled) {
            return;
        }
        this.onClick.emit(this.command);
    };
    DropdownItemDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[dropdownItem]',
                    exportAs: 'dropdownItem'
                },] }
    ];
    /** @nocollapse */
    DropdownItemDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: NavbarState }
    ]; };
    DropdownItemDirective.propDecorators = {
        isDisabled: [{ type: HostBinding, args: ['class.disabled',] }, { type: Input }],
        command: [{ type: Input, args: ['dropdownItem',] }],
        onMouseOver: [{ type: HostListener, args: ['mouseover',] }],
        onItemClick: [{ type: HostListener, args: ['click',] }]
    };
    return DropdownItemDirective;
}());
export { DropdownItemDirective };
if (false) {
    /** @type {?} */
    DropdownItemDirective.prototype.isDisabled;
    /** @type {?} */
    DropdownItemDirective.prototype.command;
    /** @type {?} */
    DropdownItemDirective.prototype.onHover;
    /** @type {?} */
    DropdownItemDirective.prototype.onClick;
    /**
     * @type {?}
     * @private
     */
    DropdownItemDirective.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    DropdownItemDirective.prototype.barState;
}
