// @farris/dynamic-control-group/esm5/lib/components/date/date-range.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from "@angular/core";
import { FormGroup } from "@angular/forms";
var DateRangeComponent = /** @class */ (function () {
    function DateRangeComponent() {
    }
    /**
     * @return {?}
     */
    DateRangeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.control.control.format = this.control.control.format || 'yyyy-MM-dd';
        this.className = this.control.className ? this.control.className : 'col-12 col-md-6 col-xl-3 col-el-2';
    };
    DateRangeComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-date-range",
                    template: "\n      <div class=\"farris-group-wrap\" [formGroup]=\"form\">\n        <div class=\"form-group farris-form-group\">\n          <label class=\"col-form-label\">\n            <span *ngIf=\"0\" class=\"farris-label-info text-danger\">*</span>\n            <span class=\"farris-label-text\">{{control.name}}</span>\n          </label>\n          <div class=\"farris-input-wrap\">\n            <farris-datepicker [dateRange]=\"true\" [returnFormat]=\"control.control.format\" [formControlName]=\"control.labelCode\" [readonly]=\"control.control.readonly\"></farris-datepicker>\n          </div>\n        </div>\n      </div>\n  "
                }] }
    ];
    /** @nocollapse */
    DateRangeComponent.ctorParameters = function () { return []; };
    DateRangeComponent.propDecorators = {
        className: [{ type: HostBinding, args: ['class',] }],
        control: [{ type: Input }],
        form: [{ type: Input }]
    };
    return DateRangeComponent;
}());
export { DateRangeComponent };
if (false) {
    /** @type {?} */
    DateRangeComponent.prototype.className;
    /** @type {?} */
    DateRangeComponent.prototype.control;
    /** @type {?} */
    DateRangeComponent.prototype.form;
}
