// @farris/dynamic-control-group/esm5/lib/components/date/single-date.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from "@angular/core";
import { FormGroup } from "@angular/forms";
var SingleDateComponent = /** @class */ (function () {
    function SingleDateComponent() {
    }
    /**
     * @return {?}
     */
    SingleDateComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.control.control.format = this.control.control.format || 'yyyy-MM-dd';
        this.className = this.control.className ? this.control.className : 'col-12 col-md-6 col-xl-3 col-el-2';
    };
    SingleDateComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-single-date",
                    template: "\n    <div class=\"farris-group-wrap\" [formGroup]=\"form\">\n      <div class=\"form-group farris-form-group\">\n        <label class=\"col-form-label\">\n          <span *ngIf=\"0\" class=\"farris-label-info text-danger\">*</span>\n          <span class=\"farris-label-text\">{{control.name}}</span>\n        </label>\n        <div class=\"farris-input-wrap\">\n          <farris-datepicker [returnFormat]=\"control.control.format\" [formControlName]=\"control.labelCode\" [readonly]=\"control.control.readonly\"></farris-datepicker>\n        </div>\n      </div>\n    </div>\n  "
                }] }
    ];
    /** @nocollapse */
    SingleDateComponent.ctorParameters = function () { return []; };
    SingleDateComponent.propDecorators = {
        className: [{ type: HostBinding, args: ['class',] }],
        control: [{ type: Input }],
        form: [{ type: Input }]
    };
    return SingleDateComponent;
}());
export { SingleDateComponent };
if (false) {
    /** @type {?} */
    SingleDateComponent.prototype.className;
    /** @type {?} */
    SingleDateComponent.prototype.control;
    /** @type {?} */
    SingleDateComponent.prototype.form;
}
