// @edp-aif/common-api/esm2015/lib/slim-entity/impl/SlimP2PMetadataEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { SlimMetadataEntity } from './SlimMetadataEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * @abstract
 */
export class SlimP2PMetadataEntity extends SlimMetadataEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.sourceBeId = null;
        this.sourceVoId = null;
        this.targetSuCode = null;
        this.targetBeId = null;
        this.targetVoId = null;
        this.targetVoProxyId = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceBeId", this.sourceBeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetSuCode", this.targetSuCode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetBeId", this.targetBeId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoId", this.targetVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyId", this.targetVoProxyId);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.sourceBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceBeId");
        this.sourceVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
        this.targetSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetSuCode");
        this.targetBeId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetBeId");
        this.targetVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoId");
        this.targetVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyId");
    }
}
if (false) {
    /** @type {?} */
    SlimP2PMetadataEntity.prototype.sourceBeId;
    /** @type {?} */
    SlimP2PMetadataEntity.prototype.sourceVoId;
    /** @type {?} */
    SlimP2PMetadataEntity.prototype.targetBeId;
    /** @type {?} */
    SlimP2PMetadataEntity.prototype.targetVoId;
    /** @type {?} */
    SlimP2PMetadataEntity.prototype.targetVoProxyId;
    /** @type {?} */
    SlimP2PMetadataEntity.prototype.targetSuCode;
}
