// @edp-aif/common-api/esm2015/lib/slim-entity/impl/SlimBizEventEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { SlimMetadataEntity } from './SlimMetadataEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
export class SlimBizEventEntity extends SlimMetadataEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.beEventRouterComponentId = null;
        this.bizEventAgentComponentId = null;
        this.voId = null;
        this.canDoExpressionId = null;
        this.bizFlowchartIdExpressionId = null;
        this.beEventRouterClassTypeName = null;
        this.isEnabled = true;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beEventRouterComponentId", this.beEventRouterComponentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventAgentComponentId", this.bizEventAgentComponentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "canDoExpressionId", this.canDoExpressionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartIdExpressionId", this.bizFlowchartIdExpressionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "beEventRouterClassTypeName", this.beEventRouterClassTypeName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.beEventRouterComponentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beEventRouterComponentId");
        this.bizEventAgentComponentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventAgentComponentId");
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
        this.canDoExpressionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "canDoExpressionId");
        this.bizFlowchartIdExpressionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartIdExpressionId");
        this.beEventRouterClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "beEventRouterClassTypeName");
    }
}
if (false) {
    /** @type {?} */
    SlimBizEventEntity.prototype.isEnabled;
    /** @type {?} */
    SlimBizEventEntity.prototype.voId;
    /** @type {?} */
    SlimBizEventEntity.prototype.beEventRouterComponentId;
    /** @type {?} */
    SlimBizEventEntity.prototype.bizEventAgentComponentId;
    /** @type {?} */
    SlimBizEventEntity.prototype.canDoExpressionId;
    /** @type {?} */
    SlimBizEventEntity.prototype.bizFlowchartIdExpressionId;
    /** @type {?} */
    SlimBizEventEntity.prototype.beEventRouterClassTypeName;
}
