// @edp-aif/common-api/esm2015/lib/slim-entity/impl/SlimMappingEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { SlimP2PMetadataEntity } from './SlimP2PMetadataEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
/**
 * @abstract
 */
export class SlimMappingEntity extends SlimP2PMetadataEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.mappingRuleMetadataId = null;
        this.mappingRuleId = null;
        this.mappingComponentId = null;
        this.mappingClassTypeName = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleMetadataId", this.mappingRuleMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingRuleId", this.mappingRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingComponentId", this.mappingComponentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingClassTypeName", this.mappingClassTypeName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.mappingRuleMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleMetadataId");
        this.mappingRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingRuleId");
        this.mappingComponentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingComponentId");
        this.mappingClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingClassTypeName");
    }
}
if (false) {
    /** @type {?} */
    SlimMappingEntity.prototype.mappingRuleMetadataId;
    /** @type {?} */
    SlimMappingEntity.prototype.mappingComponentId;
    /** @type {?} */
    SlimMappingEntity.prototype.mappingRuleId;
    /** @type {?} */
    SlimMappingEntity.prototype.mappingClassTypeName;
}
