// @edp-aif/common-api/esm2015/lib/slim-entity/impl/SlimCreationRuleEntityEx.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { SlimCreationRuleEntity } from './SlimCreationRuleEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
export class SlimCreationRuleEntityEx extends SlimCreationRuleEntity {
    /**
     * @param {?=} entity
     */
    constructor(entity) {
        super();
        if (entity != null) {
            this.id = entity.id;
            this.description = entity.description;
            this.mappingComponentId = entity.mappingComponentId;
            this.mappingRuleMetadataId = entity.mappingRuleMetadataId;
            this.name = entity.name;
            this.sourceVoId = entity.sourceVoId;
            this.sourceBeId = entity.sourceBeId;
            this.suCode = entity.suCode;
            this.targetVoId = entity.targetVoId;
            this.targetBeId = entity.targetBeId;
            this.targetSuCode = entity.targetSuCode;
            this.targetVoProxyId = entity.targetVoProxyId;
            this.formSettingsId = entity.formSettingsId;
            this.mappingRuleId = entity.mappingRuleId;
        }
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.name = null;
        this.sourceVoName = null;
        this.targetVoName = null;
        this.targetVoProxyName = null;
        this.description = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoName", this.sourceVoName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoName", this.targetVoName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyName", this.targetVoProxyName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "description", this.description);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.name = BpmnModelJsonHelper.GetAttributeValue(obj, "name");
        this.sourceVoName = BpmnModelJsonHelper.GetAttributeValue(obj, "sourceVoName");
        this.targetVoName = BpmnModelJsonHelper.GetAttributeValue(obj, "targetVoName");
        this.targetVoProxyName = BpmnModelJsonHelper.GetAttributeValue(obj, "targetVoProxyName");
        this.description = BpmnModelJsonHelper.GetAttributeValue(obj, "description");
    }
}
if (false) {
    /** @type {?} */
    SlimCreationRuleEntityEx.prototype.name;
    /** @type {?} */
    SlimCreationRuleEntityEx.prototype.sourceVoName;
    /** @type {?} */
    SlimCreationRuleEntityEx.prototype.targetVoName;
    /** @type {?} */
    SlimCreationRuleEntityEx.prototype.targetVoProxyName;
    /** @type {?} */
    SlimCreationRuleEntityEx.prototype.description;
}
