// @edp-aif/common-api/esm2015/lib/slim-entity/impl/SlimBizActionEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { SlimMetadataEntity } from './SlimMetadataEntity';
import { BizActionType } from '../../entity/BizActionType';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
export class SlimBizActionEntity extends SlimMetadataEntity {
    constructor() {
        super(...arguments);
        this.bizActionType = BizActionType.DEFAULT;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.bizActionType = BizActionType.DEFAULT;
        this.bizActionAgentComponentId = null;
        this.voId = null;
        this.bizActionAgentClassTypeName = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionType", this.bizActionType);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionAgentComponentId", this.bizActionAgentComponentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizActionAgentClassTypeName", this.bizActionAgentClassTypeName);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.bizActionType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionType");
        this.bizActionAgentComponentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionAgentComponentId");
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
        this.bizActionAgentClassTypeName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizActionAgentClassTypeName");
    }
}
if (false) {
    /** @type {?} */
    SlimBizActionEntity.prototype.bizActionType;
    /** @type {?} */
    SlimBizActionEntity.prototype.voId;
    /** @type {?} */
    SlimBizActionEntity.prototype.bizActionAgentComponentId;
    /** @type {?} */
    SlimBizActionEntity.prototype.bizActionAgentClassTypeName;
}
