// @edp-aif/common-api/esm2015/lib/slim-entity/impl/SlimAssociationEntity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { SlimEntity } from './SlimEntity';
import { BpmnModelJsonHelper } from '../../BpmnModelJsonHelper';
export class SlimAssociationEntity extends SlimEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.srcBeMetadataId = null;
        this.tarBeMetadataId = null;
        this.srcTableName = null;
        this.srcDataId = null;
        this.tarTableName = null;
        this.tarDataId = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcBeMetadataId", this.srcBeMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarBeMetadataId", this.tarBeMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcTableName", this.srcTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcDataId", this.srcDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarTableName", this.tarTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarDataId", this.tarDataId);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.srcBeMetadataId = BpmnModelJsonHelper.GetAttributeValue(obj, "srcBeMetadataId");
        this.tarBeMetadataId = BpmnModelJsonHelper.GetAttributeValue(obj, "tarBeMetadataId");
        this.srcTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcTableName");
        this.srcDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcDataId");
        this.tarTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarTableName");
        this.tarDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarDataId");
    }
}
if (false) {
    /** @type {?} */
    SlimAssociationEntity.prototype.srcBeMetadataId;
    /** @type {?} */
    SlimAssociationEntity.prototype.tarBeMetadataId;
    /** @type {?} */
    SlimAssociationEntity.prototype.srcTableName;
    /** @type {?} */
    SlimAssociationEntity.prototype.srcDataId;
    /** @type {?} */
    SlimAssociationEntity.prototype.tarTableName;
    /** @type {?} */
    SlimAssociationEntity.prototype.tarDataId;
}
