// @gsp-wf/wf-process-model/esm2015/lib/model/extension/completion-strategy/BackWardRule.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ExtensionElement, BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { TargetRefConstants } from "../../misc/TargetRefConstants";
export class BackWardRule extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = BackWardRule.Clr_Type_ID;
        this.TargetRef = TargetRefConstants.INITIAL_ACTIVITY_ID;
        this.SelectTargetAble = false;
        this.WaitReturn = false;
        this.ResubmitEffect = null;
        this.DynamicBackWardRuleAble = false;
        this.SelectTargetList = new Array();
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new BackWardRule(this.Model);
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "TargetRef", this.TargetRef);
        BpmnModelHelper.SetAttributeValue(obj, "SelectTargetAble", this.SelectTargetAble);
        BpmnModelHelper.SetAttributeValue(obj, "WaitReturn", this.WaitReturn);
        BpmnModelHelper.SetAttributeValue(obj, "ResubmitEffect", this.ResubmitEffect);
        BpmnModelHelper.SetAttributeValue(obj, "DynamicBackWardRuleAble", this.DynamicBackWardRuleAble);
        BpmnModelHelper.SetAttributeValue(obj, "SelectTargetList", this.SelectTargetList);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.TargetRef = obj.TargetRef;
        this.SelectTargetAble = obj.SelectTargetAble;
        this.WaitReturn = obj.WaitReturn;
        this.ResubmitEffect = obj.ResubmitEffect;
        this.DynamicBackWardRuleAble = obj.DynamicBackWardRuleAble;
        this.SelectTargetList = obj.SelectTargetList;
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const backWardRule = new BackWardRule(this.Model);
        backWardRule.TargetRef = this.TargetRef;
        backWardRule.SelectTargetAble = this.SelectTargetAble;
        backWardRule.WaitReturn = this.WaitReturn;
        backWardRule.ResubmitEffect = this.ResubmitEffect;
        backWardRule.DynamicBackWardRuleAble = this.DynamicBackWardRuleAble;
        backWardRule.SelectTargetList = this.SelectTargetList;
        return backWardRule;
    }
}
BackWardRule.Clr_Type_ID = "wf.BackWardRule";
if (false) {
    /** @type {?} */
    BackWardRule.Clr_Type_ID;
    /** @type {?} */
    BackWardRule.prototype.ClrTypeID;
    /** @type {?} */
    BackWardRule.prototype.TargetRef;
    /** @type {?} */
    BackWardRule.prototype.SelectTargetAble;
    /** @type {?} */
    BackWardRule.prototype.WaitReturn;
    /** @type {?} */
    BackWardRule.prototype.ResubmitEffect;
    /** @type {?} */
    BackWardRule.prototype.DynamicBackWardRuleAble;
    /** @type {?} */
    BackWardRule.prototype.SelectTargetList;
}
BpmnModelHelper.RegisterElementType(new BackWardRule(new BpmnModel()));
