// @gsp-wf/wf-process-model/esm2015/lib/model/extension/completion-strategy/CounterSignRule.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { Condition } from "../condition-expression/Condition";
import { ExtensionCompletionConditionRuler } from "@edp-pmf/bpmn-model";
export class CounterSignRule extends ExtensionCompletionConditionRuler {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = CounterSignRule.Clr_Type_ID;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new CounterSignRule(this.Model);
    }
    /**
     * 计数器
     *
     * @return {?}
     */
    get CounterVariableName() {
        return `nrOf${this.ConclusionValue}`;
    }
    /**
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "ConclusionValue", this.ConclusionValue);
        BpmnModelHelper.SetAttributeValue(obj, "CalculateType", this.CalculateType);
        BpmnModelHelper.SetAttributeValue(obj, "ConditionValue", this.ConditionValue);
        BpmnModelHelper.SetAttributeValue(obj, "WaitAllVotes", this.WaitAllVotes);
        if (this.CompletionCondition) {
            BpmnModelHelper.SetAttributeValue(obj, "CompletionCondition", this.CompletionCondition.ConvertToJsonObject());
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.ConclusionValue = obj.ConclusionValue;
        this.CalculateType = obj.CalculateType;
        this.ConditionValue = obj.ConditionValue;
        this.WaitAllVotes = obj.WaitAllVotes;
        if (obj.CompletionCondition) {
            /** @type {?} */
            let condition = new Condition(this.Model);
            condition.LoadFromJsonObject(obj.CompletionCondition);
            this.CompletionCondition = condition;
        }
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const counterSignRule = new CounterSignRule(this.Model);
        counterSignRule.ConclusionValue = this.ConclusionValue;
        counterSignRule.CalculateType = this.CalculateType;
        counterSignRule.ConditionValue = this.ConditionValue;
        counterSignRule.WaitAllVotes = this.WaitAllVotes;
        if (this.CompletionCondition) {
            counterSignRule.CompletionCondition = this.CompletionCondition.clone();
        }
        return counterSignRule;
    }
}
CounterSignRule.Clr_Type_ID = 'wf.CounterSignRule';
if (false) {
    /** @type {?} */
    CounterSignRule.Clr_Type_ID;
    /** @type {?} */
    CounterSignRule.prototype.ClrTypeID;
    /**
     * 结论值
     *
     * @type {?}
     */
    CounterSignRule.prototype.ConclusionValue;
    /**
     * 完成条件
     *
     * @type {?}
     */
    CounterSignRule.prototype.CompletionCondition;
    /**
     * 计算方式
     * @type {?}
     */
    CounterSignRule.prototype.CalculateType;
    /**
     * 满足所有条件值
     * @type {?}
     */
    CounterSignRule.prototype.ConditionValue;
    /**
     * 是否等待所有选票
     * @type {?}
     */
    CounterSignRule.prototype.WaitAllVotes;
}
BpmnModelHelper.RegisterElementType(new CounterSignRule(new BpmnModel()));
