// @gsp-wf/wf-process-model/esm2015/lib/model/extension/completion-strategy/ParticipantCoefficient.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelHelper, MexElement } from "@edp-pmf/bpmn-model";
export class ParticipantCoefficient extends MexElement {
    /**
     * @param {?} model
     * @param {?=} id
     * @param {?=} name
     * @param {?=} code
     * @param {?=} type
     * @param {?=} mustCompleted
     * @param {?=} coefficient
     */
    constructor(model, id, name, code, type, mustCompleted, coefficient) {
        super(model);
        this.participantId = id || '';
        this.participantName = name || '';
        this.participantCode = code || '';
        this.participantType = type || '';
        this.mustCompleted = mustCompleted || true;
        this.coefficient = coefficient || 0;
    }
    //序列化内容
    /**
     * @return {?}
     */
    get ParticipantId() { return this.participantId; }
    /**
     * @param {?} value
     * @return {?}
     */
    set ParticipantId(value) { this.participantId = value; }
    /**
     * @return {?}
     */
    get ParticipantName() { return this.participantName; }
    /**
     * @param {?} value
     * @return {?}
     */
    set ParticipantName(value) { this.participantName = value; }
    /**
     * @return {?}
     */
    get ParticipantCode() { return this.participantCode; }
    /**
     * @param {?} value
     * @return {?}
     */
    set ParticipantCode(value) { this.participantCode = value; }
    /**
     * @return {?}
     */
    get ParticipantType() { return this.participantType; }
    /**
     * @param {?} value
     * @return {?}
     */
    set ParticipantType(value) { this.participantType = value; }
    /**
     * @return {?}
     */
    get MustCompleted() { return this.mustCompleted; }
    /**
     * @param {?} value
     * @return {?}
     */
    set MustCompleted(value) { this.mustCompleted = value; }
    /**
     * @return {?}
     */
    get Coefficient() { return this.coefficient; }
    /**
     * @param {?} value
     * @return {?}
     */
    set Coefficient(value) { this.coefficient = value; }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, 'ParticipantId', this.ParticipantId);
        BpmnModelHelper.SetAttributeValue(obj, 'ParticipantName', this.ParticipantName);
        BpmnModelHelper.SetAttributeValue(obj, 'ParticipantCode', this.ParticipantCode);
        BpmnModelHelper.SetAttributeValue(obj, 'ParticipantType', this.ParticipantType);
        BpmnModelHelper.SetAttributeValue(obj, 'MustCompleted', this.MustCompleted);
        BpmnModelHelper.SetAttributeValue(obj, 'Coefficient', this.Coefficient);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.ParticipantId = obj.ParticipantId;
        this.ParticipantCode = obj.ParticipantCode;
        this.ParticipantName = obj.ParticipantName;
        this.ParticipantType = obj.ParticipantType;
        this.MustCompleted = obj.MustCompleted;
        this.Coefficient = obj.Coefficient;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    ParticipantCoefficient.prototype.participantId;
    /**
     * @type {?}
     * @private
     */
    ParticipantCoefficient.prototype.participantName;
    /**
     * @type {?}
     * @private
     */
    ParticipantCoefficient.prototype.participantCode;
    /**
     * @type {?}
     * @private
     */
    ParticipantCoefficient.prototype.participantType;
    /**
     * @type {?}
     * @private
     */
    ParticipantCoefficient.prototype.mustCompleted;
    /**
     * @type {?}
     * @private
     */
    ParticipantCoefficient.prototype.coefficient;
}
