// @farris/ui-datepicker/esm5/lib/constants/constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Constants
 * @type {?}
 */
var D = 'd';
/** @type {?} */
var DD = 'dd';
/** @type {?} */
var M = 'M';
/** @type {?} */
var MM = 'MM';
/** @type {?} */
var MMM = 'MMM';
/** @type {?} */
var Y = 'y';
/** @type {?} */
var YYYY = 'yyyy';
/** @type {?} */
var DATE_ROW_COUNT = 5;
/** @type {?} */
var DATE_COL_COUNT = 6;
/** @type {?} */
var MONTH_ROW_COUNT = 3;
/** @type {?} */
var MONTH_COL_COUNT = 2;
/** @type {?} */
var YEAR_ROW_COUNT = 4;
/** @type {?} */
var YEAR_COL_COUNT = 4;
/** @type {?} */
var DOT = '.';
/** @type {?} */
var UNDER_LINE = '_';
/** @type {?} */
var SU = 'su';
/** @type {?} */
var MO = 'mo';
/** @type {?} */
var TU = 'tu';
/** @type {?} */
var WE = 'we';
/** @type {?} */
var TH = 'th';
/** @type {?} */
var FR = 'fr';
/** @type {?} */
var SA = 'sa';
/** @type {?} */
var DEFAULT_LOCALE = 'ZH_CN';
/** @type {?} */
var ZERO_STR = '0';
/** @type {?} */
var EMPTY_STR = '';
/** @type {?} */
var CLICK = 'click';
/** @type {?} */
var KEYUP = 'keyup';
/** @type {?} */
var BLUR = 'blur';
/** @type {?} */
var DISABLED = 'disabled';
/** @type {?} */
var BODY = 'body';
/** @type {?} */
var VALUE = 'value';
/** @type {?} */
var OPTIONS = 'options';
/** @type {?} */
var DEFAULT_MONTH = 'defaultMonth';
/** @type {?} */
var LOCALE = 'locale';
/** @type {?} */
var OBJECT = 'object';
/** @type {?} */
var PX = 'px';
/** @type {?} */
var STYLE = 'style';
/** @type {?} */
var OPTS = 'opts';
/** @type {?} */
var YEARS = 'years';
/** @type {?} */
var VISIBLE_MONTH = 'visibleMonth';
/** @type {?} */
var SELECT_MONTH = 'selectMonth';
/** @type {?} */
var SELECT_YEAR = 'selectYear';
/** @type {?} */
var PREV_VIEW_DISABLED = 'prevViewDisabled';
/** @type {?} */
var NEXT_VIEW_DISABLED = 'nextViewDisabled';
/** @type {?} */
var DATES = 'dates';
/** @type {?} */
var WEEK_DAYS = 'weekDays';
/** @type {?} */
var SELECTED_DATE = 'selectedDate';
/** @type {?} */
var SELECTED_DATE_RANGE = 'selectedDateRange';
/** @type {?} */
var MONTHS = 'months';
/** @type {?} */
var PREVENT_CLOSE_TIMEOUT = 50;
export { D, DD, M, MM, MMM, Y, YYYY, DATE_ROW_COUNT, DATE_COL_COUNT, MONTH_ROW_COUNT, MONTH_COL_COUNT, YEAR_ROW_COUNT, YEAR_COL_COUNT, DOT, UNDER_LINE, SU, MO, TU, WE, TH, FR, SA, DEFAULT_LOCALE, ZERO_STR, EMPTY_STR, CLICK, KEYUP, BLUR, DISABLED, BODY, VALUE, OPTIONS, DEFAULT_MONTH, LOCALE, OBJECT, PX, STYLE, OPTS, YEARS, VISIBLE_MONTH, SELECT_MONTH, SELECT_YEAR, PREV_VIEW_DISABLED, NEXT_VIEW_DISABLED, DATES, WEEK_DAYS, SELECTED_DATE, SELECTED_DATE_RANGE, MONTHS, PREVENT_CLOSE_TIMEOUT };
