// @farris/ui-list-view/esm2015/lib/component/list-view-checkout.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
export class ListViewCheckoutComponent {
    constructor() {
        this.checkedChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @param {?} event
     * @return {?}
     */
    handleClick(event) {
        event.stopPropagation();
        if (!this.disabled) {
            this.checkedChange.emit({ originalEvent: event, checked: !this.checked });
        }
    }
}
ListViewCheckoutComponent.decorators = [
    { type: Component, args: [{
                selector: 'list-view-checkout',
                template: "<div class=\"custom-control custom-checkbox  custom-control-inline listview-checkbox\">\r\n    <input type=\"checkbox\" class=\"custom-control-input\"  [checked]=\"checked\">\r\n    <label class=\"custom-control-label\" (click)=\"handleClick($event)\"></label>\r\n</div>"
            }] }
];
/** @nocollapse */
ListViewCheckoutComponent.ctorParameters = () => [];
ListViewCheckoutComponent.propDecorators = {
    checked: [{ type: Input }],
    id: [{ type: Input }],
    disabled: [{ type: Input }],
    checkedChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    ListViewCheckoutComponent.prototype.checked;
    /** @type {?} */
    ListViewCheckoutComponent.prototype.id;
    /** @type {?} */
    ListViewCheckoutComponent.prototype.disabled;
    /** @type {?} */
    ListViewCheckoutComponent.prototype.checkedChange;
}
