// @gsp-wf/wf-process-editor/esm5/lib/views/time-limit-control/time-limit-control.entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ProcessDeUtil } from '../../domain/process-de-util';
var TimeLimitInfo = /** @class */ (function () {
    function TimeLimitInfo() {
        this.tag = 'constant';
        this.enable = false;
        this.timeUnit = 'day';
        this.expireOperations = new Array();
    }
    return TimeLimitInfo;
}());
export { TimeLimitInfo };
if (false) {
    /** @type {?} */
    TimeLimitInfo.prototype.tag;
    /**
     * 是否启用
     * @type {?}
     */
    TimeLimitInfo.prototype.enable;
    /**
     * 限定时长
     * @type {?}
     */
    TimeLimitInfo.prototype.dueTime;
    /**
     * 时间单位
     * @type {?}
     */
    TimeLimitInfo.prototype.timeUnit;
    /** @type {?} */
    TimeLimitInfo.prototype.expireOperations;
}
var ExpireOperation = /** @class */ (function () {
    function ExpireOperation() {
        this.id = ProcessDeUtil.GenerateElementId();
        this.durationType = 'BeforeDuration';
        this.timeDuration = 0;
        this.timeUnit = 'day';
        // 默认操作类型是发送预警消息
        this.operationTypeName = 'SendMessageExpireOperation';
    }
    return ExpireOperation;
}());
export { ExpireOperation };
if (false) {
    /** @type {?} */
    ExpireOperation.prototype.id;
    /**
     * 期限控制类型
     * @type {?}
     */
    ExpireOperation.prototype.durationType;
    /**
     * 任务期限
     * @type {?}
     */
    ExpireOperation.prototype.timeDuration;
    /**
     * 时间单位
     * @type {?}
     */
    ExpireOperation.prototype.timeUnit;
    /**
     * 操作类型名称
     * @type {?}
     */
    ExpireOperation.prototype.operationTypeName;
}
/**
 * 发送预警信息方式
 */
var /**
 * 发送预警信息方式
 */
SendMessage = /** @class */ (function (_super) {
    tslib_1.__extends(SendMessage, _super);
    function SendMessage(location, localePipe) {
        var _this = _super.call(this) || this;
        _this.localePipe = localePipe;
        _this.cycleEnable = false;
        _this.cycleTimeUnit = 'day';
        _this.receiver = location === 'activity' ? 'workItemParticipant' : 'procInstCreator';
        _this.receiverText = location === 'activity' ? _this.localePipe.transform('component.timeLimitControl.workItemParticipant') : _this.localePipe.transform('component.timeLimitControl.procInstCreator');
        _this.subject = "{\"expr\":\"'" + _this.localePipe.transform('component.timeLimitControl.SendMessageExpireOperation') + "'\",\"sexpr\":\"\"}";
        _this.message = "{\"expr\":\"'" + (location === 'activity' ? _this.localePipe.transform('component.timeLimitControl.to-doMessageTimeOut') : _this.localePipe.transform('component.timeLimitControl.messageTimeOut')) + "'\",\"sexpr\":\"\"}";
        return _this;
    }
    return SendMessage;
}(ExpireOperation));
/**
 * 发送预警信息方式
 */
export { SendMessage };
if (false) {
    /**
     * 预警信息内容
     * @type {?}
     */
    SendMessage.prototype.message;
    /**
     * 发送预警信息类型
     * @type {?}
     */
    SendMessage.prototype.sendType;
    /**
     * 预警信息主题
     * @type {?}
     */
    SendMessage.prototype.subject;
    /**
     * 消息接收人
     * 用户IDs
     * @type {?}
     */
    SendMessage.prototype.receiver;
    /**
     * 消息接收人显示
     * @type {?}
     */
    SendMessage.prototype.receiverText;
    /**
     * 启用循环预警
     * @type {?}
     */
    SendMessage.prototype.cycleEnable;
    /**
     * 循环预警时间间隔
     * @type {?}
     */
    SendMessage.prototype.cycleTime;
    /**
     * 循环预警时间间隔单位
     * @type {?}
     */
    SendMessage.prototype.cycleTimeUnit;
    /**
     * 短信模板
     * @type {?}
     */
    SendMessage.prototype.smsMessageTemplate;
    /**
     * @type {?}
     * @private
     */
    SendMessage.prototype.localePipe;
}
var AutoHandle = /** @class */ (function (_super) {
    tslib_1.__extends(AutoHandle, _super);
    function AutoHandle(type, localePipe) {
        var _this = _super.call(this) || this;
        _this.localePipe = localePipe;
        _this.autoHandleType = type;
        _this.message = _this.localePipe.transform('component.timeLimitControl.agree');
        return _this;
    }
    return AutoHandle;
}(ExpireOperation));
export { AutoHandle };
if (false) {
    /**
     * 办理类型，通过驳回否决等等
     * @type {?}
     */
    AutoHandle.prototype.autoHandleType;
    /**
     * 办理意见
     * @type {?}
     */
    AutoHandle.prototype.message;
    /**
     * @type {?}
     * @private
     */
    AutoHandle.prototype.localePipe;
}
