// @farris/extend-file-upload/esm5/lib/directive/f-file-select.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directive/f-file-select.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { FFileUploaderService } from '../service/f-file-uploader.service';
var FFileSelectDirective = /** @class */ (function () {
    function FFileSelectDirective(elementRef, upload) {
        var _this = this;
        this.elementRef = elementRef;
        this.upload = upload;
        this.tDisabled = false;
        this.changeListener = null;
        this.onChange = (/**
         * @return {?}
         */
        function () {
            if (_this.el.files) {
                _this.upload.handleFiles(_this.el.files);
                _this.el.value = null;
            }
        });
        this.uploadOutput = new EventEmitter();
        this.el = this.elementRef.nativeElement;
    }
    Object.defineProperty(FFileSelectDirective.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this.tDisabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (this.tDisabled !== value) {
                this.tDisabled = value;
                if (value) {
                    this.unbindEvent();
                }
                else {
                    this.bindEvent();
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FFileSelectDirective.prototype, "options", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value) {
                this.upload.setOptions(value);
            }
        },
        enumerable: true,
        configurable: true
    });
    ;
    Object.defineProperty(FFileSelectDirective.prototype, "uploadedCount", {
        // 已上传
        set: 
        // 已上传
        /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.upload.setOptions({ 'uploadedCount': value });
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FFileSelectDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._sub = [];
        // 对外抛出事件
        this._sub.push(this.upload.serviceEvents.subscribe((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.uploadOutput.emit(event);
        })));
        // 对内的事件
        if (this.uploadInput instanceof EventEmitter) {
            this._sub.push(this.upload.initInputEvents(this.uploadInput));
        }
        if (!this.disabled) {
            this.bindEvent();
        }
    };
    /**
     * @private
     * @return {?}
     */
    FFileSelectDirective.prototype.bindEvent = /**
     * @private
     * @return {?}
     */
    function () {
        this.changeListener = this.onChange.bind(this);
        this.el.addEventListener('change', this.changeListener);
    };
    /**
     * @private
     * @return {?}
     */
    FFileSelectDirective.prototype.unbindEvent = /**
     * @private
     * @return {?}
     */
    function () {
        this.changeListener = null;
        this.el.removeEventListener('change', this.changeListener);
    };
    /**
     * @return {?}
     */
    FFileSelectDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.el) {
            this._sub.forEach((/**
             * @param {?} sub
             * @return {?}
             */
            function (sub) { return sub.unsubscribe(); }));
        }
    };
    /**
     * 重置
     */
    /**
     * 重置
     * @return {?}
     */
    FFileSelectDirective.prototype.reset = /**
     * 重置
     * @return {?}
     */
    function () {
        // if (this.el) {
        //   //this._sub.forEach(sub => sub.unsubscribe());
        //   //this._sub=[];
        //   this.upload.reset();
        // }    
        this.upload.reset();
    };
    FFileSelectDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[FFileSelect]'
                },] }
    ];
    /** @nocollapse */
    FFileSelectDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: FFileUploaderService }
    ]; };
    FFileSelectDirective.propDecorators = {
        disabled: [{ type: Input }],
        options: [{ type: Input }],
        uploadedCount: [{ type: Input }],
        uploadInput: [{ type: Input }],
        uploadOutput: [{ type: Output }]
    };
    return FFileSelectDirective;
}());
export { FFileSelectDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FFileSelectDirective.prototype.tDisabled;
    /** @type {?} */
    FFileSelectDirective.prototype.uploadInput;
    /** @type {?} */
    FFileSelectDirective.prototype.uploadOutput;
    /** @type {?} */
    FFileSelectDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FFileSelectDirective.prototype._sub;
    /**
     * @type {?}
     * @private
     */
    FFileSelectDirective.prototype.changeListener;
    /** @type {?} */
    FFileSelectDirective.prototype.onChange;
    /** @type {?} */
    FFileSelectDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    FFileSelectDirective.prototype.upload;
    /* Skipping unhandled member: ;*/
}
