// @farris/extend-file-upload/esm5/lib/directive/f-file-drop.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directive/f-file-drop.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { FFileUploaderService } from '../service/index';
var FFileDropDirective = /** @class */ (function () {
    function FFileDropDirective(elementRef, upload) {
        this.elementRef = elementRef;
        this.upload = upload;
        this.tDisabled = false;
        // eventListener
        this.dropListener = null;
        this.dragOverListener = null;
        this.dragLeaveListener = null;
        this.dragEnterListener = null;
        this.onDropEnter = (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            e.stopPropagation();
            e.preventDefault();
        });
        this.uploadOutput = new EventEmitter();
    }
    Object.defineProperty(FFileDropDirective.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this.tDisabled;
        },
        // 是否禁用
        set: 
        // 是否禁用
        /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (this.tDisabled !== value) {
                this.tDisabled = value;
                if (value) {
                    this.unbindDragAndDropEvent();
                }
                else {
                    this.bindDragAndDropEvent();
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    ;
    Object.defineProperty(FFileDropDirective.prototype, "options", {
        // 配置
        set: 
        // 配置
        /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.upload.setOptions(value);
        },
        enumerable: true,
        configurable: true
    });
    ;
    Object.defineProperty(FFileDropDirective.prototype, "uploadedCount", {
        // 已上传
        set: 
        // 已上传
        /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.upload.setOptions({ 'uploadedCount': value });
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FFileDropDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._sub = [];
        this.el = this.elementRef.nativeElement;
        this._sub.push(this.upload.serviceEvents.subscribe((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.uploadOutput.emit(event);
        })));
        if (this.uploadInput instanceof EventEmitter) {
            this._sub.push(this.upload.initInputEvents(this.uploadInput));
        }
        // 绑定事件
        if (!this.disabled) {
            this.bindDragAndDropEvent();
        }
    };
    /**
     * @return {?}
     */
    FFileDropDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.unbindDragAndDropEvent();
        this._sub.forEach((/**
         * @param {?} sub
         * @return {?}
         */
        function (sub) { return sub.unsubscribe(); }));
    };
    /**
     * @private
     * @return {?}
     */
    FFileDropDirective.prototype.bindDragAndDropEvent = /**
     * @private
     * @return {?}
     */
    function () {
        // this.ngzone.runOutsideAngular(() => {
        this.dropListener = this.onDrop.bind(this);
        this.el.addEventListener('drop', this.dropListener);
        this.dragOverListener = this.onDragOver.bind(this);
        this.el.addEventListener('dragover', this.dragOverListener);
        this.dragLeaveListener = this.onDragLeave.bind(this);
        this.el.addEventListener('dragleave', this.dragLeaveListener);
        this.dragEnterListener = this.onDropEnter.bind(this);
        this.el.addEventListener('dragenter', this.dragEnterListener);
        //  });
    };
    /**
     * @private
     * @return {?}
     */
    FFileDropDirective.prototype.unbindDragAndDropEvent = /**
     * @private
     * @return {?}
     */
    function () {
        this.dropListener && this.el.removeEventListener('drop', this.dropListener);
        this.dragOverListener && this.el.removeEventListener('dragover', this.dragOverListener);
        this.dragLeaveListener && this.el.removeEventListener('dragleave', this.dragLeaveListener);
        this.dragEnterListener && this.el.removeEventListener('dragenter', this.dragEnterListener);
    };
    // @HostListener('drop', ['$event'])
    // @HostListener('drop', ['$event'])
    /**
     * @param {?} e
     * @return {?}
     */
    FFileDropDirective.prototype.onDrop = 
    // @HostListener('drop', ['$event'])
    /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        e.stopPropagation();
        e.preventDefault();
        /** @type {?} */
        var event = { type: 'drop' };
        // console.log('dorp Events');
        this.uploadOutput.emit(event);
        this.upload.handleFiles(e.dataTransfer.files);
    };
    // @HostListener('dragover', ['$event'])
    // @HostListener('dragover', ['$event'])
    /**
     * @param {?} e
     * @return {?}
     */
    FFileDropDirective.prototype.onDragOver = 
    // @HostListener('dragover', ['$event'])
    /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (!e) {
            return;
        }
        e.stopPropagation();
        e.preventDefault();
        /** @type {?} */
        var event = { type: 'dragOver' };
        this.uploadOutput.emit(event);
    };
    // @HostListener('dragleave', ['$event'])
    // @HostListener('dragleave', ['$event'])
    /**
     * @param {?} e
     * @return {?}
     */
    FFileDropDirective.prototype.onDragLeave = 
    // @HostListener('dragleave', ['$event'])
    /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (!e) {
            return;
        }
        e.stopPropagation();
        e.preventDefault();
        /** @type {?} */
        var event = { type: 'dragOut' };
        this.uploadOutput.emit(event);
    };
    /**
     * 重置
     */
    /**
     * 重置
     * @return {?}
     */
    FFileDropDirective.prototype.reset = /**
     * 重置
     * @return {?}
     */
    function () {
        /*if (this.el) {
             this._sub.forEach(sub => sub.unsubscribe());
             this._sub = [];
             this.upload.reset();
         }*/
        this.upload.reset();
    };
    FFileDropDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[FFileDrop]'
                },] }
    ];
    /** @nocollapse */
    FFileDropDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: FFileUploaderService }
    ]; };
    FFileDropDirective.propDecorators = {
        disabled: [{ type: Input }],
        options: [{ type: Input }],
        uploadedCount: [{ type: Input }],
        uploadInput: [{ type: Input }],
        uploadOutput: [{ type: Output }]
    };
    return FFileDropDirective;
}());
export { FFileDropDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FFileDropDirective.prototype.tDisabled;
    /** @type {?} */
    FFileDropDirective.prototype.uploadInput;
    /** @type {?} */
    FFileDropDirective.prototype.uploadOutput;
    /** @type {?} */
    FFileDropDirective.prototype.el;
    /** @type {?} */
    FFileDropDirective.prototype._sub;
    /** @type {?} */
    FFileDropDirective.prototype.dropListener;
    /** @type {?} */
    FFileDropDirective.prototype.dragOverListener;
    /** @type {?} */
    FFileDropDirective.prototype.dragLeaveListener;
    /** @type {?} */
    FFileDropDirective.prototype.dragEnterListener;
    /** @type {?} */
    FFileDropDirective.prototype.onDropEnter;
    /** @type {?} */
    FFileDropDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    FFileDropDirective.prototype.upload;
    /* Skipping unhandled member: ;*/
    /* Skipping unhandled member: ;*/
}
