// @farris/command-services/esm5/lib/manifest/form_expression_manifest_service.js
import { HttpClient } from "@angular/common/http";
import { Inject, Injectable, Injector } from "@angular/core";
import { of } from "rxjs";
import { share, switchMap } from "rxjs/operators";
import { FORM_PATH_TOKEN, FrameContext, FORM_MANIFEST_SERVICE_TOKEN } from "@farris/devkit";
import { BasePathService } from "@farris/rtf";
var FormExpressionManifestService = /** @class */ (function () {
    function FormExpressionManifestService(injector, frameContext, httpClient, modulePath, formManifestService) {
        this.injector = injector;
        this.frameContext = frameContext;
        this.httpClient = httpClient;
        this.modulePath = modulePath;
        this.formManifestService = formManifestService;
    }
    FormExpressionManifestService.prototype.load = function () {
        var _this = this;
        if (this.modulePath.endsWith('\/')) {
            this.modulePath = this.modulePath.substring(0, this.modulePath.length - 1);
        }
        return this.formManifestService.load().pipe(switchMap(function (formManifest) {
            var expressions = formManifest.expressions;
            var expressionManifest = expressions.find(function (expressionManifest) { return expressionManifest.ns === _this.frameContext.namespace; });
            if (expressionManifest) {
                var expressionPath = BasePathService.convertPath(_this.modulePath + "/expressions/" + expressionManifest.path + "?version=" + new Date().valueOf().toString());
                var key = BasePathService.convertPath(_this.modulePath + "/expressions/" + expressionManifest.path);
                var request$ = FormExpressionManifestService.mainfests.get(key);
                if (request$) {
                    return request$;
                }
                else {
                    var req$ = _this.httpClient.get(expressionPath, { responseType: 'json' }).pipe(share());
                    FormExpressionManifestService.mainfests.set(key, req$);
                    return req$;
                }
            }
            else {
                return of({});
            }
        }));
    };
    FormExpressionManifestService.mainfests = new Map();
    FormExpressionManifestService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FormExpressionManifestService.ctorParameters = function () { return [
        { type: Injector },
        { type: FrameContext },
        { type: HttpClient },
        { type: undefined, decorators: [{ type: Inject, args: [FORM_PATH_TOKEN,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [FORM_MANIFEST_SERVICE_TOKEN,] }] }
    ]; };
    return FormExpressionManifestService;
}());
export { FormExpressionManifestService };
