// @farris/ui-wizard/esm5/lib/model/wizard.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var WizardEventType = {
    Finish: 0,
    Cancel: 1,
    Prev: 2,
    Next: 3,
    StepChange: 4,
    ChangeIndex: 5,
};
export { WizardEventType };
WizardEventType[WizardEventType.Finish] = 'Finish';
WizardEventType[WizardEventType.Cancel] = 'Cancel';
WizardEventType[WizardEventType.Prev] = 'Prev';
WizardEventType[WizardEventType.Next] = 'Next';
WizardEventType[WizardEventType.StepChange] = 'StepChange';
WizardEventType[WizardEventType.ChangeIndex] = 'ChangeIndex';
var WizardEvent = /** @class */ (function () {
    function WizardEvent(state, fromPageId, toPageId) {
        this.state = state;
        this.fromPageId = fromPageId;
        this.toPageId = toPageId;
    }
    return WizardEvent;
}());
export { WizardEvent };
if (false) {
    /** @type {?} */
    WizardEvent.prototype.state;
    /** @type {?} */
    WizardEvent.prototype.fromPageId;
    /** @type {?} */
    WizardEvent.prototype.toPageId;
}
