// @farris/mobile-devkit/esm2015/lib/state-machine/state_machine.js
import { BehaviorSubject } from 'rxjs';
import { MetadataUtil } from '../core/index';
import { State, initialUIState } from './types';
import { STATE_PROP_META, RENDER_STATE_PROP_META, ACTION_METHOD_META } from './decorators';
import { StateMachineContext } from './state_machine_context';
import { StateMachineWatcher } from './state_machine_watcher';
/**
 * 状态机
 */
export class StateMachine {
    /**
     * 构造函数
     */
    constructor() {
        this.renderStates = {};
        this.handlePropMetadatas();
        this.stateChange = new BehaviorSubject(false);
        this.context = new StateMachineContext(this, this.initialState);
        this.stateMachineWatcher = new StateMachineWatcher(this);
    }
    /**
     * 初始化状态机
     * @param viewModelContext ViewModel上下文
     * @summary
     * 状态机变更，为了在绑定数据之后执行状态机的操作，把render方法延后执行。
     */
    init(viewModelContext) {
        this.viewModelContext = viewModelContext;
        this.context.init(this.viewModelContext);
        this.stateMachineWatcher.init(this.viewModelContext);
        this.render();
    }
    /**
     * 批量处理属性元数据
     */
    handlePropMetadatas() {
        const propsMetadatas = MetadataUtil.getPropsMetadatas(this.constructor);
        // 遍历所有属性装饰器，并调用相应的build方法
        if (propsMetadatas) {
            Object.keys(propsMetadatas).forEach((propName) => {
                const propMetadatas = propsMetadatas[propName];
                propMetadatas.forEach(propMetadata => {
                    this.handlePropMetadata(propName, propMetadata);
                });
            });
        }
        if (!this.initialState) {
            throw new Error('请在StatePropMeta注解中指定状态机的初始状态。');
        }
    }
    /**
     * 处理属性元数据
     */
    handlePropMetadata(propName, propMetadata) {
        const ngMetadataName = propMetadata.ngMetadataName;
        switch (ngMetadataName) {
            case STATE_PROP_META:
                this.buildState(propName, propMetadata);
                break;
            case RENDER_STATE_PROP_META:
                this.buildRenderState(propName, propMetadata);
                break;
            case ACTION_METHOD_META:
                this.buildAction(propName, propMetadata);
                break;
            default:
                break;
        }
    }
    /**
     * 包装State
     * @param stateName 状态名称
     * @param ngState   状态对象
     */
    buildState(stateName, ngState) {
        this.states = this.states || {};
        this[stateName] = new State(stateName);
        this.states[stateName] = this[stateName];
        if (ngState.initialState) {
            this.initialState = this[stateName];
        }
    }
    /**
     * 包装RenderState
     * @param renderStateName 渲染状态名称
     * @param ngRenderState   渲染状态元数据
     */
    buildRenderState(renderStateName, ngRenderState) {
        this.renderStates = this.renderStates || {};
        this[renderStateName] = initialUIState;
        this.renderStates[renderStateName] = this[renderStateName];
        // 将renderState上指定的render加入到renders中
        this.renders = this.renders || {};
        this.renders[renderStateName] = ngRenderState.render;
    }
    /**
     * 包装Action
     * @param actionName 动作名称
     * @param ngAction 动作元数据
     */
    buildAction(actionName, ngAction) {
        this[actionName] = () => {
            const nextStateName = ngAction.transitTo;
            const nextState = this.states[nextStateName];
            this.context.transitTo(nextState.name);
            this.render();
        };
    }
    /**
     * 重新计算所有渲染状态的值
     * @sumamry
     * 当 state切换的时候，调用遍历所有的render方法，更改renderState
     */
    render() {
        for (const renderStateName in this.renderStates) {
            if (this.renderStates.hasOwnProperty(renderStateName) === false) {
                continue;
            }
            // 执行RenderState的render方法，更新renderState
            const stateRender = this.renders[renderStateName];
            if (!stateRender) {
                continue;
            }
            this.renderStates[renderStateName] = stateRender(this.context);
            this[renderStateName] = this.renderStates[renderStateName];
        }
        this.stateChange.next(this.context.state);
    }
}
