// @gsp-cmp/common-component/esm2015/convertor/data-type/structure-refer-type-convertor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class StructureReferTypeConvertor {
    /**
     * @param {?} obj
     * @param {?} datatype
     * @param {?} ctxt
     * @return {?}
     */
    convertToObject(obj, datatype, ctxt) {
        /** @type {?} */
        var referType = (/** @type {?} */ (datatype));
        obj["kind"] = referType.getKind();
        obj['referId'] = referType.referId;
        obj['code'] = referType.code;
        obj['name'] = referType.name;
    }
    /**
     * @param {?} dataType
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    initFromObject(dataType, obj, ctxt) {
        /** @type {?} */
        var refType = (/** @type {?} */ (dataType));
        refType.kind = obj['kind'];
        refType.referId = obj['referId'];
        refType.code = obj['code'];
        refType.name = obj['name'];
    }
}
