// @gsp-wf/wf-process-editor/esm5/lib/views/urge-handle-rule/urge-handle-rule.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var UrgeHandleRule = /** @class */ (function () {
    function UrgeHandleRule() {
    }
    /**
     * @return {?}
     */
    UrgeHandleRule.prototype.cconstructor = /**
     * @return {?}
     */
    function () {
        this.urgeHandleAble = false;
        this.earlyTime = 0;
        this.intervalTime = 0;
        this.urgeMessageSubject = "{\"expr\":\"'催办审批：' + DefaultFunction.GetContextParameter(\\\"bizDefName\\\")\",\"sexpr\":\"\"}";
        this.urgeMessageContent = "{\"expr\":\"'请您尽快审批' + DefaultFunction.GetContextParameter(\\\"procInstCreatorName\\\") + '的' + DefaultFunction.GetContextParameter(\\\"bizDefName\\\") + '。'\",\"sexpr\":\"\"}";
        this.addReminder = false;
        this.subjectEditable = true;
        this.contentEditable = true;
    };
    return UrgeHandleRule;
}());
export { UrgeHandleRule };
if (false) {
    /** @type {?} */
    UrgeHandleRule.prototype.urgeHandleAble;
    /** @type {?} */
    UrgeHandleRule.prototype.sendType;
    /** @type {?} */
    UrgeHandleRule.prototype.earlyTime;
    /** @type {?} */
    UrgeHandleRule.prototype.intervalTime;
    /** @type {?} */
    UrgeHandleRule.prototype.urgeMessageSubject;
    /** @type {?} */
    UrgeHandleRule.prototype.urgeMessageContent;
    /** @type {?} */
    UrgeHandleRule.prototype.addReminder;
    /** @type {?} */
    UrgeHandleRule.prototype.subjectEditable;
    /** @type {?} */
    UrgeHandleRule.prototype.contentEditable;
    /**
     * 短信模板
     * @type {?}
     */
    UrgeHandleRule.prototype.smsMessageTemplate;
}
