// @farris/mobile-command-services/esm2015/lib/ccwork-service/ccwork.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ccwork-service/ccwork.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class CcWorkService {
    // public location$ = new Subject();
    constructor() {
    }
    /**
     * 云上协同关闭页面
     * @return {?}
     */
    closeCcWorkWebView() {
        window['iCityBridge'].closeWebView();
    }
    /**
     * 云上协同隐藏标题栏
     * @return {?}
     */
    hideCcWorkNaviBar() {
        window['iCityBridge'].hideTitleBar();
    }
    /**
     *  云上协同更改标题栏
     * @param {?} title
     * @return {?}
     */
    setTitle(title) {
        window['iCityBridge'].setTitle(title);
    }
    /**
     * 获取当前位置
     * @param {?} callBack
     * @return {?}
     */
    getLocation(callBack) {
        if (window['iCityBridge']) {
            window['iCityBridge'].ccworkGetLocation(callBack);
        }
    }
    /**
     * 获取二维码
     * @param {?} callBack
     * @param {?=} params 配置项
     * @return {?}
     */
    scanCode(callBack, params = {}) {
        if (window['iCityBridge']) {
            /** @type {?} */
            const paramsObj = Object.assign({ filters: [
                    'QR', 'EAN13', 'EAN8'
                ], path: '' }, params);
            window['iCityBridge'].ccworkScanCode(paramsObj, callBack);
        }
    }
    /**
     * 云上协同调用拨号
     * @param {?} phoneNumber 手机号
     * @param {?} callback
     * @return {?}
     */
    callPhone(phoneNumber, callback) {
        if (window['iCityBridge']) {
            /** @type {?} */
            const params = {
                number: phoneNumber
            };
            window['iCityBridge'].ccworkCallPhone(params, callback);
        }
    }
    /**
     * 从相机和相册中获取照片
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    getPicsFromCameraAndAlbum(params, callback) {
        if (window['iCityBridge']) {
            window['iCityBridge'].ccworkGetPicsFromCameraAndAlbum(params, callback);
        }
    }
}
