// @gsp-wf/wf-dimension-def/esm5/lib/wf-dimension-def.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { FarrisButtonModule } from '@farris/ui-button';
import { MessagerModule } from '@farris/ui-messager';
import { DatagridEditorsModule, EditorProviders } from '@farris/ui-datagrid-editors';
import { DatagridModule } from '@farris/ui-datagrid';
import { InputGroupModule } from '@farris/ui-input-group';
import { NotifyModule } from '@farris/ui-notify';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { ComboListModule } from '@farris/ui-combo-list';
import { WfDimensionDefComponent } from './wf-dimension-def.component';
import { MetadataRTSelectModule, GSPMetadataRTServiceModule, GSPMetadataRTService } from '@gsp-lcm/metadatart-selector';
import { DimensionItemCardComponent } from './dimension-item-card/dimension-item-card.component';
import { HttpService } from '@ecp-caf/caf-common';
import { WFDimensionDefViewmodel } from './service/wf-dimension-def.viewModel';
import { WFDimensionDefUIState } from './service/wf-dimension-def.uistate';
import { ExpressionModule, ExpressionService } from '@gsp-svc/expression';
import { EnumEditComponent } from './enum-edit/enum-edit.component';
import { WFDimensionPanelModule } from './refModule/panel/panel.module';
import { SelectElementModule } from './refModule/select-element/select-element.module';
import { LookupConfigModule } from '@farris/lookup-config';
var WfDimensionDefModule = /** @class */ (function () {
    function WfDimensionDefModule() {
    }
    WfDimensionDefModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        WfDimensionDefComponent,
                        DimensionItemCardComponent,
                        EnumEditComponent
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        FarrisButtonModule,
                        MessagerModule.forRoot(),
                        FarrisDialogModule,
                        NotifyModule,
                        GSPMetadataRTServiceModule.forRoot(''),
                        MetadataRTSelectModule,
                        InputGroupModule,
                        WFDimensionPanelModule,
                        DatagridEditorsModule,
                        DatagridModule.forRoot(tslib_1.__spread(EditorProviders)),
                        ComboListModule,
                        ExpressionModule,
                        SelectElementModule,
                        LookupConfigModule
                    ],
                    providers: [
                        GSPMetadataRTService,
                        WFDimensionDefViewmodel,
                        WFDimensionDefUIState,
                        ExpressionService,
                        HttpService
                    ],
                    entryComponents: [
                        EnumEditComponent
                    ],
                    exports: [WfDimensionDefComponent, DimensionItemCardComponent]
                },] }
    ];
    return WfDimensionDefModule;
}());
export { WfDimensionDefModule };
