// @gsp-wf/wf-dimension-def/esm5/lib/wf-dimension-def.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { MessagerService } from '@farris/ui-messager';
import { DimenDefEntityUtil } from './entity/dimendef-entity.util';
import { ItemEditState, DimensionInputType } from './entity/dimensiondef.entity';
import { WFDimensionDefViewmodel } from './service/wf-dimension-def.viewModel';
import { NotifyService } from '@farris/ui-notify';
import { WFDimensionDefUIState } from './service/wf-dimension-def.uistate';
import { forkJoin } from 'rxjs/observable/forkJoin';
import { switchMap, catchError } from 'rxjs/operators';
import { of } from 'rxjs';
var WfDimensionDefComponent = /** @class */ (function () {
    function WfDimensionDefComponent(messageService, viewmodel, notifyService, uistate) {
        this.messageService = messageService;
        this.viewmodel = viewmodel;
        this.notifyService = notifyService;
        this.uistate = uistate;
        this.entry = 'flowForm';
        // tslint:disable-next-line: variable-name
        this._editable = false;
        this.data = new Array();
        this.editData = new Array();
    }
    Object.defineProperty(WfDimensionDefComponent.prototype, "flowFormId", {
        get: /**
         * @return {?}
         */
        function () {
            return this._flowFormId;
        },
        set: /**
         * @param {?} id
         * @return {?}
         */
        function (id) {
            this._flowFormId = id;
            this.load();
            this.getDimensionDefSet(this._flowFormId);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WfDimensionDefComponent.prototype, "schema", {
        get: /**
         * @return {?}
         */
        function () {
            return this.uistate.schema;
        },
        set: /**
         * @param {?} s
         * @return {?}
         */
        function (s) {
            this.uistate.schema = s;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WfDimensionDefComponent.prototype, "beMetaData", {
        get: /**
         * @return {?}
         */
        function () {
            return this.uistate.beMetaData;
        },
        set: /**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            this.uistate.beMetaData = m;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WfDimensionDefComponent.prototype, "editable", {
        get: /**
         * @return {?}
         */
        function () {
            return this._editable;
        },
        set: /**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            var e_1, _a;
            this._editable = e;
            try {
                for (var _b = tslib_1.__values(this.editData), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var d = _c.value;
                    d.canEdit = this._editable;
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    WfDimensionDefComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    WfDimensionDefComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
    };
    /**
     * @return {?}
     */
    WfDimensionDefComponent.prototype.load = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.flowFormId) {
            this.viewmodel.loadDimensionDefs(this.flowFormId).subscribe((/**
             * @param {?} re
             * @return {?}
             */
            function (re) {
                _this.data = re;
                _this.editData = [];
                if (_this.data && _this.data.length > 0) {
                    _this.data.forEach((/**
                     * @param {?} item
                     * @return {?}
                     */
                    function (item) {
                        _this.addItem(item, _this.editable);
                    }));
                }
            }));
        }
    };
    /**
     * @param {?} flowFormId
     * @return {?}
     */
    WfDimensionDefComponent.prototype.getDimensionDefSet = /**
     * @param {?} flowFormId
     * @return {?}
     */
    function (flowFormId) {
        var _this = this;
        if (flowFormId) {
            this.viewmodel.getDimensionDefSet(flowFormId).subscribe((/**
             * @param {?} re
             * @return {?}
             */
            function (re) {
                _this.dimensionDefSet = re;
            }));
        }
    };
    /**
     * @param {?} index
     * @return {?}
     */
    WfDimensionDefComponent.prototype.hasItem = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        return this.editData.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.index === index; }));
    };
    /**
     * @param {?} index
     * @return {?}
     */
    WfDimensionDefComponent.prototype.getItem = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        return this.hasItem(index) ? this.editData.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.index === index; })).itemData : null;
    };
    /**
     * @param {?} index
     * @param {?} itemData
     * @return {?}
     */
    WfDimensionDefComponent.prototype.setItem = /**
     * @param {?} index
     * @param {?} itemData
     * @return {?}
     */
    function (index, itemData) {
        this.editData.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.index === index; })).itemData = itemData;
    };
    /**
     * @param {?} itemData
     * @param {?} canEdit
     * @return {?}
     */
    WfDimensionDefComponent.prototype.addItem = /**
     * @param {?} itemData
     * @param {?} canEdit
     * @return {?}
     */
    function (itemData, canEdit) {
        /** @type {?} */
        var editDataItem = new ItemEditState();
        editDataItem.canEdit = canEdit;
        editDataItem.index = this.editData.length;
        editDataItem.itemData = Object.assign({}, itemData);
        editDataItem.id = itemData.id;
        this.editData.push(editDataItem);
    };
    /**
     * @param {?} index
     * @return {?}
     */
    WfDimensionDefComponent.prototype.getTitle = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        /** @type {?} */
        var suffix = '';
        switch (index) {
            case 0:
                suffix = '一';
                break;
            case 1:
                suffix = '二';
                break;
        }
        return "\u7EF4\u5EA6" + suffix;
    };
    /**
     * @param {?} index
     * @return {?}
     */
    WfDimensionDefComponent.prototype.getState = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        if (!this.hasItem(index)) {
            return false;
        }
        /** @type {?} */
        var sysInit = this.editData.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.index === index; })).itemData.sysInit;
        return sysInit === 1 ? true : false;
    };
    /**
     * @param {?} index
     * @return {?}
     */
    WfDimensionDefComponent.prototype.isEditState = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        if (!this.hasItem(index)) {
            return false;
        }
        return this.editData.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.index === index; })).canEdit;
    };
    /**
     * @param {?} item
     * @return {?}
     */
    WfDimensionDefComponent.prototype.edit = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.editData.find((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return d.id === item.id; })).canEdit = true;
    };
    /**
     * @param {?} item
     * @return {?}
     */
    WfDimensionDefComponent.prototype.remove = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        var _this = this;
        if (this.dimensionDefSet && (item.id === this.dimensionDefSet.dimension1 || item.id === this.dimensionDefSet.dimension2)) {
            this.messageService.info('当前维度定义已在流程分配维度配置中启用');
            return;
        }
        if (this.entry === 'flowForm') {
            this.removeDimensionDef(item);
            // const i = this.data.findIndex(d => d.id === item.id);
            // if (i > -1) {
            //   this.data.splice(i, 1);
            // }
        }
        else {
            this.messageService.question("\u786E\u8BA4\u5220\u9664'" + this.getTitle(item.index) + "'\u5417\uFF1F", (/**
             * @return {?}
             */
            function () {
                if (_this.data.findIndex((/**
                 * @param {?} d
                 * @return {?}
                 */
                function (d) { return d.id === item.id; })) === -1) {
                    _this.removeDimensionDef(item);
                    _this.notifyService.success('删除成功');
                    return;
                }
                _this.viewmodel.removeDimensionDef(item.id).subscribe((/**
                 * @param {?} re
                 * @return {?}
                 */
                function (re) {
                    _this.removeDimensionDef(item);
                    /** @type {?} */
                    var i = _this.data.findIndex((/**
                     * @param {?} d
                     * @return {?}
                     */
                    function (d) { return d.id === item.id; }));
                    if (i > -1) {
                        _this.data.splice(i, 1);
                    }
                    _this.notifyService.success('删除成功');
                }), (/**
                 * @param {?} err
                 * @return {?}
                 */
                function (err) {
                    _this.notifyService.error('删除失败，请联系管理员查看原因！');
                }));
            }));
        }
    };
    /**
     * @return {?}
     */
    WfDimensionDefComponent.prototype.getDimensionDefs = /**
     * @return {?}
     */
    function () {
        return this.editData.map((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return d.itemData; }));
    };
    /**
     * @private
     * @param {?} item
     * @return {?}
     */
    WfDimensionDefComponent.prototype.removeDimensionDef = /**
     * @private
     * @param {?} item
     * @return {?}
     */
    function (item) {
        /** @type {?} */
        var index = this.editData.findIndex((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return d.id === item.id; }));
        if (index > -1) {
            this.editData.splice(index, 1);
        }
        if (index < this.editData.length) {
            for (var i = 0; i < this.editData.length; i++) {
                this.editData[i].index = i;
            }
        }
    };
    /**
     * @param {?} item
     * @return {?}
     */
    WfDimensionDefComponent.prototype.save = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        var _this = this;
        /** @type {?} */
        var v = this.validate(item.itemData);
        if (!v.result) {
            this.messageService.info(v.message);
            return;
        }
        if (this.data.findIndex((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return d.id === item.id; })) === -1) {
            this.viewmodel.addDimensionDef(item.itemData).subscribe((/**
             * @param {?} re
             * @return {?}
             */
            function (re) {
                _this.endEdit(item);
                _this.data.push(Object.assign({}, item.itemData));
                _this.notifyService.success('保存成功');
            }), (/**
             * @param {?} err
             * @return {?}
             */
            function (err) {
                _this.notifyService.error('保存失败，请联系管理员查看原因！');
            }));
        }
        else {
            this.viewmodel.updateDimensionDef(item.itemData).subscribe((/**
             * @param {?} re
             * @return {?}
             */
            function (re) {
                _this.endEdit(item);
                /** @type {?} */
                var data = _this.data.find((/**
                 * @param {?} d
                 * @return {?}
                 */
                function (d) { return d.id === item.id; }));
                Object.assign(data, item.itemData);
                _this.notifyService.success('保存成功');
            }), (/**
             * @param {?} err
             * @return {?}
             */
            function (err) {
                _this.notifyService.error('保存失败，请联系管理员查看原因！');
            }));
        }
    };
    /**
     * @return {?}
     */
    WfDimensionDefComponent.prototype.saveAll = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var deleteDefs = this.data.filter((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return _this.editData.findIndex((/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return e.id === d.id; })) === -1; }));
        /** @type {?} */
        var addDefs = this.editData.filter((/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return _this.data.findIndex((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return d.id === e.id; })) === -1; })).map((/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return e.itemData; }));
        /** @type {?} */
        var updateDefs = this.editData.filter((/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return _this.data.findIndex((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return d.id === e.id; })) > -1; }))
            .map((/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return e.itemData; }))
            .filter((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return d.sysInit === 0; }));
        return forkJoin(deleteDefs.map((/**
         * @param {?} d
         * @return {?}
         */
        function (d) {
            return _this.viewmodel.removeDimensionDef(d.id);
        })).concat(updateDefs.map((/**
         * @param {?} u
         * @return {?}
         */
        function (u) {
            return _this.viewmodel.updateDimensionDef(u);
        })).concat(addDefs.map((/**
         * @param {?} a
         * @return {?}
         */
        function (a) {
            return _this.viewmodel.addDimensionDef(a);
        }))))).pipe(switchMap((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            _this.data = _this.editData.map((/**
             * @param {?} e
             * @return {?}
             */
            function (e) { return e.itemData; }));
            // this.editData.map(e => e.canEdit = false);
            return of(true);
        })), catchError((/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            console.error(err);
            return of(false);
        })));
        //   .subscribe(re => {
        //   this.data = this.editData.map(e => e.itemData);
        //   this.editData.map(e => e.canEdit = false);
        // },
        // err => {
        //   this.notifyService.error('保存维度定义失败，请联系系统管理员查看原因！');
        // });
    };
    /**
     * @param {?} item
     * @return {?}
     */
    WfDimensionDefComponent.prototype.cancel = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.endEdit(item);
        if (this.data.findIndex((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return d.id === item.id; })) === -1) {
            this.removeDimensionDef(item);
            return;
        }
        else {
            Object.assign(this.editData.find((/**
             * @param {?} d
             * @return {?}
             */
            function (d) {
                return d.id === item.id;
            })).itemData, this.data.find((/**
             * @param {?} d
             * @return {?}
             */
            function (d) { return d.id === item.id; })));
        }
    };
    /**
     * @return {?}
     */
    WfDimensionDefComponent.prototype.addHandler = /**
     * @return {?}
     */
    function () {
        if (this.entry === 'flowForm' && !this.editable) {
            return;
        }
        /** @type {?} */
        var info = DimenDefEntityUtil.newDimendefEntity(this.flowFormId);
        this.addItem(info, true);
    };
    /**
     * @param {?} index
     * @return {?}
     */
    WfDimensionDefComponent.prototype.isOtherItemEditting = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        var _this = this;
        /** @type {?} */
        var isOtherItemEditting = false;
        this.editData.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (item.canEdit === true && item.index !== index) {
                isOtherItemEditting = true;
                _this.messageService.info("\u8BF7\u5148\u4FDD\u5B58\u5BF9\u4E8E'" + _this.getTitle(item.index) + "'\u7684\u4FEE\u6539\u3002");
            }
        }));
        return isOtherItemEditting;
    };
    /**
     * @param {?} item
     * @return {?}
     */
    WfDimensionDefComponent.prototype.endEdit = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        var e_2, _a;
        try {
            for (var _b = tslib_1.__values(this.editData), _c = _b.next(); !_c.done; _c = _b.next()) {
                var d = _c.value;
                if (d.id === item.id) {
                    d.canEdit = false;
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    /**
     * @private
     * @param {?} d
     * @return {?}
     */
    WfDimensionDefComponent.prototype.validate = /**
     * @private
     * @param {?} d
     * @return {?}
     */
    function (d) {
        if (!d.code) {
            return { result: false, message: '维度定义不允许编号为空，请修改！' };
        }
        else if (!d.dimensionName) {
            return { result: false, message: '维度定义不允许维度名称为空，请修改！' };
        }
        else if (!d.dimensionCode) {
            return { result: false, message: '维度定义不允许分配维度为空，请修改！' };
        }
        else if (d.inputType === DimensionInputType.Help && !d.smartHelp) {
            return { result: false, message: '维度定义数据类型为帮助，但帮助值为空，请修改！' };
        }
        else if (d.inputType === DimensionInputType.Enum && !d.enumList) {
            return { result: false, message: '维度定义数据类型为枚举，但枚举值为空，请修改！' };
        }
        return { result: true, message: '' };
    };
    WfDimensionDefComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-dimension-def',
                    template: "<wf-dimension-panel [model]=\"['b']\" [width]=\"'700px'\" [defaultExpand]=\"true\">\r\n    <div *ngFor=\"let item of editData; index as index;\"\r\n        style=\"width: 100%;border: 1px solid rgba(0, 0, 0, .125);margin-bottom: 25px;\">\r\n        <wf-dimension-panel-item [title]=\"getTitle(index)\" [sysInit]=\"getState(index)\" [value]=\"index\" [editState]=\"entry==='billCategory'?isEditState(index):false\">\r\n            <ng-template #contentTempl>\r\n                <wf-dimension-item-card [data]=\"getItem(index)\" [canCardEdit]=\"isEditState(index)\" [sysInit]=\"getState(index)\">\r\n                </wf-dimension-item-card>\r\n            </ng-template>\r\n            <ng-template #toolTempl>\r\n                <farris-button *ngIf=\"entry==='billCategory'\" [type]=\"'link'\" (click)=\"edit(item)\"><span class=\"f-icon text-info k-i-edit\"></span>\u4FEE\u6539\r\n                </farris-button>\r\n                <farris-button *ngIf=\"editable && !getState(index)\" [type]=\"'link'\" (click)=\"remove(item)\"><span\r\n                        class=\"f-icon text-info k-i-delete\"></span> \u5220\u9664\r\n                </farris-button>\r\n            </ng-template>\r\n            <ng-template #bottomRef>\r\n                <farris-button (click)=\"save(item)\">\u4FDD\u5B58</farris-button>\r\n                <farris-button style=\"margin-left:10px\" [type]=\"'secondary'\" (click)=\"cancel(item)\">\u53D6\u6D88</farris-button>\r\n            </ng-template>\r\n        </wf-dimension-panel-item>\r\n    </div>\r\n</wf-dimension-panel>\r\n<div *ngIf=\"hasItem(1)\" class=\"panel-ony-two\">\r\n    <div style=\"width: 250px;margin:0 auto\">\r\n        <p style=\"color:rgba(217,217,217)\">Tips:\u76EE\u524D\u4EC5\u652F\u6301\u6DFB\u52A0\u4E24\u4E2A\u7EF4\u5EA6</p>\r\n    </div>\r\n</div>\r\n<div *ngIf=\"!hasItem(1)\" class=\"panel-add\">\r\n    <div style=\"width: 100px;margin:0 auto\">\r\n        <farris-button style=\"margin:25px\" [type]=\"'link'\" [disabled]=\"entry==='flowForm'&&!editable\" (click)=\"addHandler()\">\r\n            <span *ngIf=\"!(entry==='flowForm'&&!editable)\" class=\"f-icon text-info k-i-plus\"></span>\u6DFB\u52A0\u7EF4\u5EA6</farris-button>\r\n    </div>\r\n</div>",
                    styles: [".panel-add{margin:25px;padding:10px;border:1px dashed rgba(0,0,0,.125)}.panel-ony-two{margin:20px}"]
                }] }
    ];
    /** @nocollapse */
    WfDimensionDefComponent.ctorParameters = function () { return [
        { type: MessagerService },
        { type: WFDimensionDefViewmodel },
        { type: NotifyService },
        { type: WFDimensionDefUIState }
    ]; };
    WfDimensionDefComponent.propDecorators = {
        flowFormId: [{ type: Input }],
        schema: [{ type: Input }],
        beMetaData: [{ type: Input }],
        entry: [{ type: Input }],
        editable: [{ type: Input }]
    };
    return WfDimensionDefComponent;
}());
export { WfDimensionDefComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    WfDimensionDefComponent.prototype._flowFormId;
    /** @type {?} */
    WfDimensionDefComponent.prototype.entry;
    /** @type {?} */
    WfDimensionDefComponent.prototype._editable;
    /**
     * @type {?}
     * @private
     */
    WfDimensionDefComponent.prototype.data;
    /** @type {?} */
    WfDimensionDefComponent.prototype.editData;
    /** @type {?} */
    WfDimensionDefComponent.prototype.dimensionDefSet;
    /**
     * @type {?}
     * @private
     */
    WfDimensionDefComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    WfDimensionDefComponent.prototype.viewmodel;
    /**
     * @type {?}
     * @private
     */
    WfDimensionDefComponent.prototype.notifyService;
    /**
     * @type {?}
     * @private
     */
    WfDimensionDefComponent.prototype.uistate;
}
