// @farris/discussion-group/esm5/lib/discussion-list/discussion-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, ViewEncapsulation } from '@angular/core';
import { PaginationControlsComponent } from '@farris/ui-pagination';
import { DiscussionListService } from './discussion-list.service';
var DiscussionListComponent = /** @class */ (function () {
    function DiscussionListComponent(discussionListService) {
        this.discussionListService = discussionListService;
        // @Input() personnelsPrimaryKey: string = 'userId';
        this.personnelsDisplayKey = 'userName';
        this.pagerOnServer = true;
        /* *是否支持分页 */
        this.supportPaging = true;
        /**  总页数 */
        // @Input() pageSize:number;
        /**
         * 当前页码
         */
        this.pageIndex = 1;
        /**  每页显示条数选项 */
        // private _pageList = [10,20,30,50];
        // get pageList() {
        //     return this._pageList;
        // }
        // @Input() set pageList(val) {
        //     this._pageList = val;
        //     if (this.pager) {
        //         this.pager.setPageList(val);
        //     }
        // }
        /**
         * 总条数
         */
        this._total = 0;
        /**
         * 每页显示个数
         */
        this._pageSize = 10;
        /**  总条数 */
        // @Input() total = 0;
        /**
         * 评论数据
         */
        this.innerDiscussionData = [];
        this._discussionData = [];
        this.replyMessage = new EventEmitter();
        // 分页事件
        this.pageChanged = new EventEmitter();
        this.pageSizeChanged = new EventEmitter();
        /**
         * 附件预览
         */
        this.filePreview = new EventEmitter();
        /**
         * 附件下载
         */
        this.fileDownload = new EventEmitter();
        this.labels = {
            previousLabel: ' ',
            nextLabel: ' '
        };
        this.directionLinks = true;
        this.maxSize = 7;
        this.responsive = true;
        this.autoHide = false;
    }
    Object.defineProperty(DiscussionListComponent.prototype, "total", {
        get: /**
         * @return {?}
         */
        function () {
            return this._total;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._total = val;
            // this.paginationOptions.totalItems = val;
            this.initPaginationOptions();
            // if(val){
            //   this.initPaginationOptions();
            // }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DiscussionListComponent.prototype, "pageSize", {
        get: /**
         * @return {?}
         */
        function () {
            return this._pageSize;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._pageSize = val;
            // this.paginationOptions.itemsPerPage = val;
            this.initPaginationOptions();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DiscussionListComponent.prototype, "discussionData", {
        get: /**
         * @return {?}
         */
        function () {
            return this._discussionData;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val) {
                this._discussionData = val;
                this.innerDiscussionData = val;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DiscussionListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // this.paginationOptions = {
        //   id: 'Farris-discussion-Pagination',
        //   itemsPerPage: this.pageSize,
        //   currentPage: this.pageIndex,
        //   // pageList: this.pageList,
        //   totalItems: this.total
        // };
        this.initPaginationOptions();
        /** 监听指令得到的数据 */
        this.discussionListService.getData().subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data && !_this._discussionData.length) {
                _this.innerDiscussionData = data;
            }
        }));
    };
    /**
     * @return {?}
     */
    DiscussionListComponent.prototype.initPaginationOptions = /**
     * @return {?}
     */
    function () {
        this.paginationOptions = {
            id: 'Farris-discussion-Pagination',
            itemsPerPage: this._pageSize,
            currentPage: this.pageIndex,
            pageList: [10, 20, 30, 50, 100],
            totalItems: this._total,
            remote: this.pagerOnServer
        };
    };
    /** 点击回复留言 */
    /**
     * 点击回复留言
     * @param {?} item
     * @return {?}
     */
    DiscussionListComponent.prototype.reply = /**
     * 点击回复留言
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.replyMessage.emit(item);
    };
    /** 页码变化 */
    /**
     * 页码变化
     * @param {?} page
     * @return {?}
     */
    DiscussionListComponent.prototype.onPageChange = /**
     * 页码变化
     * @param {?} page
     * @return {?}
     */
    function (page) {
        if (this.pageIndex !== page.pageIndex) {
            this.pageIndex = page.pageIndex;
            this.paginationOptions.currentPage = page.pageIndex;
            this.pageChanged.emit({ pageInfo: page });
        }
    };
    /** 每页显示条数变化 */
    /**
     * 每页显示条数变化
     * @param {?} pageSize
     * @return {?}
     */
    DiscussionListComponent.prototype.onPageSizeChange = /**
     * 每页显示条数变化
     * @param {?} pageSize
     * @return {?}
     */
    function (pageSize) {
        if (this.pageSize !== pageSize && this.total) {
            this.paginationOptions.itemsPerPage = pageSize;
            this.pageSize = pageSize;
            /** @type {?} */
            var total = this.total;
            /** @type {?} */
            var pageLength = Math.floor(total / pageSize);
            if (total % pageSize > 0) {
                pageLength += 1;
            }
            if (pageLength && this.pageIndex > pageLength) {
                this.pageIndex = pageLength;
                this.paginationOptions.currentPage = this.pageIndex;
            }
            this.pageSizeChanged.emit({ pageInfo: { pageIndex: this.pageIndex, pageSize: pageSize } });
        }
    };
    /** 附件预览 */
    /**
     * 附件预览
     * @param {?} info
     * @return {?}
     */
    DiscussionListComponent.prototype.filePreviewEventHandler = /**
     * 附件预览
     * @param {?} info
     * @return {?}
     */
    function (info) {
        this.filePreview.emit(info);
    };
    /** 附件下载 */
    /**
     * 附件下载
     * @param {?} info
     * @return {?}
     */
    DiscussionListComponent.prototype.fileDownloadEventHandler = /**
     * 附件下载
     * @param {?} info
     * @return {?}
     */
    function (info) {
        this.fileDownload.emit(info);
    };
    /** 占位头像文字 */
    /**
     * 占位头像文字
     * @param {?} item
     * @return {?}
     */
    DiscussionListComponent.prototype.getAvatar = /**
     * 占位头像文字
     * @param {?} item
     * @return {?}
     */
    function (item) {
        if (item && item[this.personnelsDisplayKey]) {
            /** @type {?} */
            var str = item[this.personnelsDisplayKey];
            return str.substring(str.length - 2, str.length);
        }
        else {
            return '';
        }
    };
    DiscussionListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-discussion-list',
                    template: "<div class=\"f-discussion-group-content\">\r\n    <div class=\"f-discussion-group-content-item\" *ngFor=\"let item of (supportPaging ? (innerDiscussionData | paginate: paginationOptions) : innerDiscussionData)\">\r\n        <div class=\"discussion-item-avatar\">\r\n            <img [src]=\"item.imgData\" alt=\"\" class=\"discussion-item-avatar-img\" *ngIf=\"item.imgData\">\r\n            <div class=\"discussion-item-avatar-tip\" *ngIf=\"!item.imgData\" [innerHTML]=\"getAvatar(item)\"></div>\r\n        </div>\r\n        <div class=\"discussion-item-inner\">\r\n            <div class=\"discussion-item-username\">\r\n                {{item[personnelsDisplayKey]}}\r\n            </div>\r\n            <div class=\"discussion-item-text\">\r\n                <span class=\"discussion-item-text-message\" [innerHTML]=\"item.text | safeHtml\"></span>\r\n                <!-- <span class=\"discussion-item-text-about\" *ngIf=\"item.messageAbout\">@{{item.messageAbout}}</span> -->\r\n            </div>\r\n            <div class=\"discussion-item-text-reply\" *ngIf=\"item.parentData\">\r\n                <span class=\"discussion-item-text-reply-title\">\r\n                    {{ 'discussionGroup.reply' | locale }}\r\n                    <span class=\"discussion-item-text-reply-name\">{{item.parentData[personnelsDisplayKey]}}</span>\uFF1A\r\n                </span>\r\n                <span class=\"discussion-item-text-reply-content\" [innerHTML]=\"item.parentData.text | safeHtml\">\r\n                </span>\r\n            </div>\r\n            <!-- \u9644\u4EF6\u5217\u8868 <div class=\"discussion-item-files\" *ngIf=\"item.attachFiles && item.attachFiles.length\">\r\n                <f-file-preview class=\"d-flex flex-wrap\" [itemCls]=\"'demo-view-item'\" [fileInfos]=\"item.attachFiles\"\r\n                    (filePreviewEvent)=\"filePreviewEventHandler($event)\" (fileDownloadEvent)=\"fileDownloadEventHandler($event)\"\r\n                    #filePreview>\r\n                    <ng-template filePreviewListTypeTemplate let-data let-readonly=\"readonly\">\r\n                        <div class=\"ffilepreview--item-detail\">\r\n                            <div class=\"ffilepreview--item-icon\">\r\n                                <span class=\"ffilepreview--filetype-icon\" [ngClass]=\"filePreview.getFileTypeClassName(data.name)\"></span>\r\n                            </div>\r\n                            <div class=\"ffilepreview--item-content\">\r\n                                <a class=\"item-content--title\" [title]=\"data.name\"\r\n                                    (click)=\"filePreview.filePreviewHandler($event,data)\">{{data.name}}</a>\r\n                                <div class=\"item-content--toolbar f-text-light\">\r\n                                    <button class=\"btn btn-link\" (click)=\"filePreview.filePreviewHandler($event,data)\">\r\n                                    <span class=\"f-icon f-icon-preview\"></span>\u9884\u89C8\r\n                                    </button>\r\n                                    <button class=\"btn btn-link\" (click)=\"filePreview.fileDownloadHandler($event,data)\">\r\n                                    <span class=\"f-icon f-icon-download\"></span>\u4E0B\u8F7D\r\n                                    </button>\r\n                                </div>\r\n                            </div>\r\n                        </div>\r\n                    </ng-template>\r\n                </f-file-preview>\r\n            </div> -->\r\n            <div class=\"discussion-item-footer\">\r\n                <div class=\"discussion-item-time\">\r\n                    <span class=\"discussion-item-time-text\">{{item.commentDate | dateformat:'yyyy-MM-dd HH:mm'}}</span>\r\n                </div>\r\n                <div class=\"discussion-item-btns\">\r\n                    <span class=\"discussion-item-btns-start\" (click)=\"reply(item)\">\r\n                        <span class=\"f-icon f-icon-message\"></span>\r\n                        <span class=\"discussion-item-btns-start-text\">\r\n                            {{ 'discussionGroup.reply' | locale }}\r\n                        </span>\r\n                    </span>\r\n                    <!-- <span class=\"discussion-item-btns-new\">\r\n                        <span class=\"f-icon f-icon-star-outline\"></span>\r\n                    </span> -->\r\n                </div>\r\n            </div>\r\n        </div>\r\n    </div>\r\n    <div class=\"f-discussion-group-footer\" *ngIf=\"supportPaging\">\r\n        <pagination-controls #pager \r\n            [id]=\"paginationOptions.id\" \r\n            [maxSize]=\"maxSize\" \r\n            [directionLinks]=\"directionLinks\"\r\n            [autoHide]=\"autoHide\" \r\n            [responsive]=\"responsive\" \r\n            [previousLabel]=\"labels.previousLabel\"\r\n            [nextLabel]=\"labels.nextLabel\"\r\n            (pageChange)=\"onPageChange($event)\"\r\n            (pageSizeChange)=\"onPageSizeChange($event)\"\r\n            [showPageList]=\"false\"\r\n        >\r\n        </pagination-controls>\r\n    </div>\r\n</div>",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".f-discussion-group-content-item{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;margin-bottom:14px}.f-discussion-group-content-item .discussion-item-avatar{width:32px;height:32px;margin-right:10px;overflow:hidden;cursor:pointer;flex-shrink:0}.f-discussion-group-content-item .discussion-item-avatar-img{display:block;width:32px;height:32px;border-radius:50%}.f-discussion-group-content-item .discussion-item-avatar-tip{width:32px;height:32px;font-size:12px;color:#fff;text-align:center;line-height:32px;border-radius:50%;background-color:#4796ff}.f-discussion-group-content-item .discussion-item-inner{flex-shrink:1;-webkit-box-flex:1;flex-grow:1;flex-basis:0;padding-bottom:14px;border-bottom:1px solid #e3e3e3;overflow:hidden}.f-discussion-group-content-item .discussion-item-inner .discussion-item-username{margin-bottom:4px;font-size:16px;color:rgba(0,0,0,.85)}.f-discussion-group-content-item .discussion-item-inner .discussion-item-text{font-size:14px;color:rgba(0,0,0,.75)}.f-discussion-group-content-item .discussion-item-inner .discussion-item-text .discussion-item-text-message{word-break:break-all}.f-discussion-group-content-item .discussion-item-inner .discussion-item-text-reply{margin-top:11px;padding:6px 8px;font-size:0;background:#f5f5f5;border-radius:2px}.f-discussion-group-content-item .discussion-item-inner .discussion-item-text-reply-content,.f-discussion-group-content-item .discussion-item-inner .discussion-item-text-reply-title{font-size:14px;color:rgba(0,0,0,.75);line-height:20px}.f-discussion-group-content-item .discussion-item-inner .discussion-item-text-reply-title .discussion-item-text-reply-name{color:#3487e3}.f-discussion-group-content-item .discussion-item-inner .discussion-item-text-reply-content{word-break:break-all}.f-discussion-group-content-item .discussion-item-inner .discussion-item-files{margin-top:11px}.f-discussion-group-content-item .discussion-item-inner .discussion-item-files .ffilepreview--filetype-icon{width:40px;height:40px}.f-discussion-group-content-item .discussion-item-inner .discussion-item-footer{margin-top:10px;padding-right:20px;display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;-webkit-box-align:center;align-items:center;-webkit-box-pack:justify;justify-content:space-between}.f-discussion-group-content-item .discussion-item-inner .discussion-item-footer .discussion-item-time{font-size:14px;color:#b1b6c2}.f-discussion-group-content-item .discussion-item-inner .discussion-item-footer .discussion-item-btns{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center}.f-discussion-group-content-item .discussion-item-inner .discussion-item-footer .discussion-item-btns .discussion-item-btns-start{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;position:relative;cursor:pointer}.f-discussion-group-content-item .discussion-item-inner .discussion-item-footer .discussion-item-btns .discussion-item-btns-start .f-icon{color:rgba(0,0,0,.45);font-size:16px}.f-discussion-group-content-item .discussion-item-inner .discussion-item-footer .discussion-item-btns .discussion-item-btns-start .f-icon::before{vertical-align:bottom}.f-discussion-group-content-item .discussion-item-inner .discussion-item-footer .discussion-item-btns .discussion-item-btns-start .discussion-item-btns-start-text{margin-left:14px;font-size:14px;color:rgba(0,0,0,.45)}.f-discussion-group-content-item .discussion-item-inner .discussion-item-footer .discussion-item-btns .discussion-item-btns-start:hover .discussion-item-btns-start-text,.f-discussion-group-content-item .discussion-item-inner .discussion-item-footer .discussion-item-btns .discussion-item-btns-start:hover .f-icon{color:#3487e3}.f-discussion-group-content-item:last-child .discussion-item-inner{border-bottom:none}"]
                }] }
    ];
    /** @nocollapse */
    DiscussionListComponent.ctorParameters = function () { return [
        { type: DiscussionListService }
    ]; };
    DiscussionListComponent.propDecorators = {
        personnelsDisplayKey: [{ type: Input }],
        pagerOnServer: [{ type: Input }],
        supportPaging: [{ type: Input }],
        pageIndex: [{ type: Input }],
        total: [{ type: Input }],
        pageSize: [{ type: Input }],
        discussionData: [{ type: Input }],
        replyMessage: [{ type: Output }],
        pageChanged: [{ type: Output }],
        pageSizeChanged: [{ type: Output }],
        filePreview: [{ type: Output }],
        fileDownload: [{ type: Output }],
        pager: [{ type: ViewChild, args: ['pager',] }]
    };
    return DiscussionListComponent;
}());
export { DiscussionListComponent };
if (false) {
    /** @type {?} */
    DiscussionListComponent.prototype.personnelsDisplayKey;
    /** @type {?} */
    DiscussionListComponent.prototype.pagerOnServer;
    /** @type {?} */
    DiscussionListComponent.prototype.supportPaging;
    /**
     * 当前页码
     * @type {?}
     */
    DiscussionListComponent.prototype.pageIndex;
    /**
     * 总条数
     * @type {?}
     * @private
     */
    DiscussionListComponent.prototype._total;
    /**
     * 每页显示个数
     * @type {?}
     * @private
     */
    DiscussionListComponent.prototype._pageSize;
    /**
     * 评论数据
     * @type {?}
     */
    DiscussionListComponent.prototype.innerDiscussionData;
    /** @type {?} */
    DiscussionListComponent.prototype._discussionData;
    /** @type {?} */
    DiscussionListComponent.prototype.replyMessage;
    /** @type {?} */
    DiscussionListComponent.prototype.pageChanged;
    /** @type {?} */
    DiscussionListComponent.prototype.pageSizeChanged;
    /**
     * 附件预览
     * @type {?}
     */
    DiscussionListComponent.prototype.filePreview;
    /**
     * 附件下载
     * @type {?}
     */
    DiscussionListComponent.prototype.fileDownload;
    /** @type {?} */
    DiscussionListComponent.prototype.labels;
    /** @type {?} */
    DiscussionListComponent.prototype.directionLinks;
    /** @type {?} */
    DiscussionListComponent.prototype.maxSize;
    /** @type {?} */
    DiscussionListComponent.prototype.responsive;
    /** @type {?} */
    DiscussionListComponent.prototype.autoHide;
    /** @type {?} */
    DiscussionListComponent.prototype.paginationOptions;
    /** @type {?} */
    DiscussionListComponent.prototype.pager;
    /**
     * @type {?}
     * @private
     */
    DiscussionListComponent.prototype.discussionListService;
}
