// @farris/ui-datagrid/esm5/lib/services/state.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function RowDataChanges() { }
/**
 * @record
 */
export function FarrisDatagridState() { }
if (false) {
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.width;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.height;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.rowHeight;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.idField;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.data;
    /**
     * 当前页数据
     * @type {?|undefined}
     */
    FarrisDatagridState.prototype.currentPageData;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.originalData;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.showHeader;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.headerHeight;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.pageIndex;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.pageSize;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.pagerHeight;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.pagination;
    /** @type {?} */
    FarrisDatagridState.prototype.columns;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.flatColumns;
    /** @type {?} */
    FarrisDatagridState.prototype.fitColumns;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.autoFitColumns;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.showLineNumber;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.showCheckbox;
    /** @type {?} */
    FarrisDatagridState.prototype.checkboxColumnWidth;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.multiSelect;
    /**
     * 当启用多选时，点击行选中，只允许且只有一行被选中。
     * @type {?}
     */
    FarrisDatagridState.prototype.onlySelectSelf;
    /** @type {?} */
    FarrisDatagridState.prototype.selectOnCheck;
    /** @type {?} */
    FarrisDatagridState.prototype.checkOnSelect;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.selections;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.checkedRows;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.currentRow;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.currentCell;
    /** @type {?} */
    FarrisDatagridState.prototype.columnsGroup;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.virtualized;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.showFooter;
    /**
     * 是否启用异步虚拟加载数据: false - 客户端内存中取数；true - 远端服务器加载数据, 默认值为 false
     * @type {?|undefined}
     */
    FarrisDatagridState.prototype.virtualizedAsyncLoad;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.virtual;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.sortName;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.sortOrder;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.multiSort;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.remoteSort;
    /**
     * 服务器端过滤
     * @type {?|undefined}
     */
    FarrisDatagridState.prototype.remoteFilter;
    /**
     * 过滤条件
     * @type {?|undefined}
     */
    FarrisDatagridState.prototype.conditions;
    /** @type {?} */
    FarrisDatagridState.prototype.changes;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.groupRows;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.groupField;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.groupFooter;
    /**
     * 显示分组列
     * @type {?|undefined}
     */
    FarrisDatagridState.prototype.showGroupColumn;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.mergeCell;
    /** @type {?|undefined} */
    FarrisDatagridState.prototype.enableClientSortForMergeCell;
    /**
     * 存储组件初始时的相关属性设置
     * @type {?|undefined}
     */
    FarrisDatagridState.prototype.initialOptions;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * @record
 */
export function VirtualizedState() { }
if (false) {
    /**
     * 虚拟加载且不显示分页条时，记录加载新记录时的索引值
     * @type {?}
     */
    VirtualizedState.prototype.rowIndex;
    /**
     * 数据显示的索引值
     * @type {?|undefined}
     */
    VirtualizedState.prototype.startIndex;
    /** @type {?|undefined} */
    VirtualizedState.prototype.scrollTop;
    /** @type {?|undefined} */
    VirtualizedState.prototype.virtualRows;
    /** @type {?|undefined} */
    VirtualizedState.prototype.topHideHeight;
    /** @type {?|undefined} */
    VirtualizedState.prototype.bottomHideHeight;
    /**
     * 页脚数据
     * @type {?|undefined}
     */
    VirtualizedState.prototype.footer;
    /**
     * 是否由滚动条触发
     * @type {?|undefined}
     */
    VirtualizedState.prototype.isScrollChange;
}
/**
 * @record
 */
export function SelectedRow() { }
if (false) {
    /** @type {?} */
    SelectedRow.prototype.index;
    /** @type {?} */
    SelectedRow.prototype.id;
    /** @type {?} */
    SelectedRow.prototype.data;
    /**
     * 中间位置行引用
     * @type {?|undefined}
     */
    SelectedRow.prototype.dr;
    /**
     * 左侧位置行引用
     * @type {?|undefined}
     */
    SelectedRow.prototype.drl;
    /**
     * 右侧位置行引用
     * @type {?|undefined}
     */
    SelectedRow.prototype.drr;
    /** @type {?|undefined} */
    SelectedRow.prototype.editors;
}
/** @type {?} */
export var initDataGridState = {
    idField: 'id',
    showLineNumber: false,
    selections: [],
    checkedRows: [],
    currentRow: null,
    columnsGroup: null,
    columns: [],
    fitColumns: false,
    onlySelectSelf: true,
    selectOnCheck: true,
    checkOnSelect: true,
    pageIndex: 1,
    pageSize: 20,
    pagerHeight: 40,
    checkboxColumnWidth: 36,
    pagination: false,
    virtual: {
        rowIndex: 0,
        scrollTop: 0,
        topHideHeight: 0,
        bottomHideHeight: 0,
        isScrollChange: false
    },
    sortName: undefined,
    sortOrder: undefined,
    multiSort: false,
    remoteSort: true,
    changes: null,
    showGroupColumn: false
};
/**
 * @record
 */
export function CellInfo() { }
if (false) {
    /** @type {?|undefined} */
    CellInfo.prototype.isEditing;
    /** @type {?|undefined} */
    CellInfo.prototype.rowData;
    /** @type {?|undefined} */
    CellInfo.prototype.rowId;
    /** @type {?|undefined} */
    CellInfo.prototype.rowIndex;
    /** @type {?|undefined} */
    CellInfo.prototype.field;
    /** @type {?|undefined} */
    CellInfo.prototype.cellElement;
    /** @type {?|undefined} */
    CellInfo.prototype.cellRef;
    /** @type {?|undefined} */
    CellInfo.prototype.editor;
    /** @type {?|undefined} */
    CellInfo.prototype.column;
    /** @type {?|undefined} */
    CellInfo.prototype.cellEditorRef;
    /** @type {?|undefined} */
    CellInfo.prototype.lastRow;
}
/**
 * @record
 */
export function DataResult() { }
if (false) {
    /** @type {?|undefined} */
    DataResult.prototype.footer;
    /** @type {?} */
    DataResult.prototype.items;
    /** @type {?|undefined} */
    DataResult.prototype.total;
    /** @type {?|undefined} */
    DataResult.prototype.pageIndex;
    /** @type {?|undefined} */
    DataResult.prototype.pageSize;
}
/**
 * 行索引
 * @type {?}
 */
export var ROW_INDEX_FIELD = '__row_index__';
/**
 * 是否为分组行
 * @type {?}
 */
export var IS_GROUP_ROW_FIELD = '__group__';
/**
 * 引用的分组行属性
 * @type {?}
 */
export var GROUP_ROW_FIELD = '__group_parent_id__';
/**
 * 是否为分组合计行
 * @type {?}
 */
export var IS_GROUP_FOOTER_ROW_FIELD = '__group_footer__';
/** @type {?} */
export var GROUP_VISIBLE_FIELD = '__group_visible__';
/** @type {?} */
export var GROUP_LEVEL_FIELD = '__group_level__';
/**
 * @record
 */
export function DatagridGroupRow() { }
if (false) {
    /* Skipping unnamed member:
    [IS_GROUP_ROW_FIELD]: boolean;*/
    /**
     * 分组等级
     * @type {?}
     */
    DatagridGroupRow.prototype.level;
    /**
     * 所在行数据
     * @type {?}
     */
    DatagridGroupRow.prototype.data;
    /**
     * 分组字段值
     * @type {?}
     */
    DatagridGroupRow.prototype.value;
    /* Skipping unnamed member:
    [GROUP_VISIBLE_FIELD]: boolean;*/
    /**
     * 是否展开分组行
     * @type {?}
     */
    DatagridGroupRow.prototype.expanded;
    /**
     * 分组字段
     * @type {?}
     */
    DatagridGroupRow.prototype.field;
    /** @type {?} */
    DatagridGroupRow.prototype.colspan;
    /**
     * 分组总记录数
     * @type {?}
     */
    DatagridGroupRow.prototype.total;
    /**
     * 分组数据行
     * @type {?}
     */
    DatagridGroupRow.prototype.rows;
    /* Skipping unnamed member:
    [GROUP_ROW_FIELD]: any;*/
}
/** @type {?} */
export var IgnoreProperties = [
    'afterEdit',
    'app',
    'beforeCheck',
    'beforeClickCell',
    'beforeEdit',
    'beforeSelect',
    'beforeSortColumn',
    'beforeUncheck',
    'beforeUnselect',
    'beginEdit',
    'cancelEdited',
    'cd',
    'cellClick',
    'checkAll',
    'checkValuesChange',
    'checked',
    'checkedChange',
    'clientFilterService',
    'colFormatSer',
    'columnMoved',
    'columnSorted',
    'componentType',
    'dataChange',
    'dblclickRow',
    'destroy$',
    'dfs',
    'dgColumns',
    'dgContainer',
    'dgPager',
    'dgs',
    'domSanitizer',
    'editors',
    'el',
    'endEdit',
    'farrisInstances',
    'filterChanged',
    'footerTemplate',
    'groupFieldChange',
    'groupFields$',
    'inject',
    'lastCellEnterHandle',
    'loadSuccess',
    'localeService',
    'longTextArea',
    'rowAppended',
    'rowClick',
    'rowDeleted',
    'render2',
    'pageSizeChanged',
    'onDocumentMouseDown',
    'pageChanged',
    'resizeProxy',
    'resizeProxyBg',
    'ro',
    'rowUpdated',
    'scrollInstance',
    'scrollY',
    'scrollYLoad',
    'selectAll',
    'selectChanged',
    'sms',
    'selectValueChange',
    'selectValuesChange',
    'settingService',
    'smartFilterService',
    'subscriptions',
    'trackByRows',
    'unCheckAll',
    'unChecked',
    'unSelect',
    'unSelectAll',
    'zone',
    '__component_type__',
    '__pagination',
    '_active_datagrid_cls',
    '_autoFitColumns',
    '_blured_datagrid_cls',
    '_columns',
    '_custom_datagrid_cls',
    '_editable',
    '_fields',
    '_fit',
    '_fitColumns',
    '_footerData',
    '_groupField',
    '_isResized',
    '_loading',
    '_lockPagination',
    '_selectedRow',
    '_showRowGroupPanel',
    '_sizeType',
    '_total',
];
