// @gsp-lcm/dbo-selector/esm5/lib/dbo-select/dbo-select.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, TemplateRef } from '@angular/core';
import { DboHelpDevService } from '../dbo-help-dev.service';
var DboSelectComponent = /** @class */ (function () {
    function DboSelectComponent(dboHelpService) {
        this.dboHelpService = dboHelpService;
        this.url = '';
        /**
         * 帮助标题
         */
        this.title = '帮助';
        /**
         * 输出事件
         */
        this.selectedAfterOk = new EventEmitter();
        this.modalClosed = new EventEmitter();
        // 模式
        this.selectableSettings = {
            checkboxOnly: true,
            mode: 'single',
            enabled: true
        };
    }
    /**输入属性变化后 */
    /**
     * 输入属性变化后
     * @param {?} changes
     * @return {?}
     */
    DboSelectComponent.prototype.ngOnChanges = /**
     * 输入属性变化后
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
    };
    /**
     * @return {?}
     */
    DboSelectComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // 服务端获取数据
        this.dboHelpService.getAllDatabaseObject(this.url)
            .subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            _this.data = result;
            _this.curSelectedKeys = [];
            if (_this.initSelectedKeys) {
                _this.initSelectedKeys.map((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return _this.curSelectedKeys.push(item); }));
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            console.log('获取数据源失败！');
            console.log(error);
        }));
    };
    /**帮助确定按钮 */
    /**
     * 帮助确定按钮
     * @return {?}
     */
    DboSelectComponent.prototype.btnOk = /**
     * 帮助确定按钮
     * @return {?}
     */
    function () {
        var _this = this;
        // 根据选中行获取实体数据
        /** @type {?} */
        var selectPosEntity = [];
        this.curSelectedKeys.forEach((/**
         * @param {?} dboID
         * @return {?}
         */
        function (dboID) {
            /** @type {?} */
            var selectDBO = _this.data.find((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item['id'] === dboID; }));
            selectPosEntity.push(selectDBO);
        }));
        if (!selectPosEntity) {
            alert('请选择数据库对象');
            return;
        }
        else {
            // 发射当前选中数据
            this.selectedAfterOk.next(selectPosEntity);
        }
        this.hideModal();
    };
    /**
     * @return {?}
     */
    DboSelectComponent.prototype.hideModal = /**
     * @return {?}
     */
    function () {
        this.modalClosed.emit();
    };
    DboSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-dbo-select',
                    template: "<div class=\"modal-body\">\r\n<div class=\"templateBlock\">\r\n  <kendo-grid [class]=\"'flex-auto'\" [selectable]=\"selectableSettings\" [kendoGridSelectBy]=\"'id'\" [selectedKeys]=\"curSelectedKeys\"\r\n    [kendoGridBinding]=\"data\">\r\n    <kendo-grid-checkbox-column width=\"10%\"></kendo-grid-checkbox-column>\r\n    <kendo-grid-column field=\"code\" title=\"\u7F16\u53F7\" width=\"40%\"></kendo-grid-column>\r\n    <kendo-grid-column field=\"name\" title=\"\u540D\u79F0\" width=\"45%\"></kendo-grid-column>\r\n  </kendo-grid>\r\n  </div>\r\n</div>\r\n<ng-template #btns>\r\n<button type=\"button\" class=\"btn btn-default btn-primary\" (click)=\"btnOk()\">\u786E\u5B9A</button>\r\n<button type=\"button\" class=\"btn btn-default\" (click)=\"hideModal()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    DboSelectComponent.ctorParameters = function () { return [
        { type: DboHelpDevService }
    ]; };
    DboSelectComponent.propDecorators = {
        url: [{ type: Input }],
        title: [{ type: Input }],
        isShow: [{ type: Input }],
        initSelectedKeys: [{ type: Input }],
        selectedAfterOk: [{ type: Output }],
        modalClosed: [{ type: Output }],
        modalFooter: [{ type: ViewChild, args: ['modalFooter',] }],
        btns: [{ type: ViewChild, args: ['btns',] }]
    };
    return DboSelectComponent;
}());
export { DboSelectComponent };
if (false) {
    /** @type {?} */
    DboSelectComponent.prototype.url;
    /**
     * 帮助标题
     * @type {?}
     */
    DboSelectComponent.prototype.title;
    /**
     * 帮助打开状态
     * @type {?}
     */
    DboSelectComponent.prototype.isShow;
    /**
     * 选中列表
     * @type {?}
     */
    DboSelectComponent.prototype.initSelectedKeys;
    /**
     * 输出事件
     * @type {?}
     */
    DboSelectComponent.prototype.selectedAfterOk;
    /** @type {?} */
    DboSelectComponent.prototype.modalClosed;
    /** @type {?} */
    DboSelectComponent.prototype.modalFooter;
    /** @type {?} */
    DboSelectComponent.prototype.btns;
    /**
     * 当前界面选中的内码
     * @type {?}
     */
    DboSelectComponent.prototype.curSelectedKeys;
    /** @type {?} */
    DboSelectComponent.prototype.selectableSettings;
    /** @type {?} */
    DboSelectComponent.prototype.data;
    /**
     * @type {?}
     * @private
     */
    DboSelectComponent.prototype.dboHelpService;
}
