// @farris/ui-tag/esm2015/lib/tag.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ElementRef, HostBinding } from '@angular/core';
export class FarrisTagComponent {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        /* 类型样式 */
        this.type = 'secondary';
        /* close emit事件 */
        this.close = new EventEmitter();
        this.exist = true;
        /* tag广播事件 */
        this.selectValue = new EventEmitter();
    }
    /* 标签大小 */
    /**
     * @return {?}
     */
    get size() {
        if (this.nySize === 'small') {
            return 'sm';
        }
        else if (this.nySize === 'large') {
            return 'lg';
        }
        else {
            return 'md';
        }
    }
    /**
     * @param {?} size
     * @return {?}
     */
    set size(size) {
        this.nySize = size;
    }
    /**
     * @return {?}
     */
    tagClass() {
        return true;
    }
    /* tag 是否选中 */
    /**
     * @return {?}
     */
    get active() {
        return this.nyActive;
    }
    ;
    /**
     * @param {?} active
     * @return {?}
     */
    set active(active) {
        this.nyActive = active;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    getValue() {
        if (this.disabled) {
            return;
        }
        this.nyActive = !this.nyActive;
        this.selectValue.emit({
            value: this.value,
            active: this.active
        });
    }
    /**
     * @return {?}
     */
    hasParent() {
        return this.el.nativeElement.parentElement.classList.contains('farris-tag');
    }
    /**
     * @param {?} e
     * @return {?}
     */
    closeHandler(e) {
        e.preventDefault();
        this.exist = false;
        this.close.emit(this.value);
    }
}
FarrisTagComponent.decorators = [
    { type: Component, args: [{
                selector: 'farris-tag',
                template: `
    <span [class]="'btn btn-'+ type +' btn-'+ size"
    [class.active]="active"
    [class.disabled]="disabled"
    [class.farris-tag-disabled]="disabled"
    (click)="getValue()" *ngIf="exist">
      <ng-content></ng-content>
      &nbsp;
      <i class="k-icon k-i-close" *ngIf="closable" (click)="closeHandler($event)"></i>
    </span>
  `,
                styles: [".k-icon.k-i-close{font-size:12px}.k-icon.k-i-close:hover{cursor:pointer}:host .btn{margin-left:5px}:host .farris-tag-disabled{cursor:text}"]
            }] }
];
/** @nocollapse */
FarrisTagComponent.ctorParameters = () => [
    { type: ElementRef }
];
FarrisTagComponent.propDecorators = {
    size: [{ type: Input }],
    type: [{ type: Input }],
    text: [{ type: Input }],
    value: [{ type: Input }],
    disabled: [{ type: Input }],
    closable: [{ type: Input }],
    close: [{ type: Output }],
    tagClass: [{ type: HostBinding, args: ['class.farris-tag',] }],
    selectValue: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FarrisTagComponent.prototype.type;
    /** @type {?} */
    FarrisTagComponent.prototype.text;
    /** @type {?} */
    FarrisTagComponent.prototype.value;
    /** @type {?} */
    FarrisTagComponent.prototype.disabled;
    /** @type {?} */
    FarrisTagComponent.prototype.closable;
    /** @type {?} */
    FarrisTagComponent.prototype.close;
    /** @type {?} */
    FarrisTagComponent.prototype.exist;
    /** @type {?} */
    FarrisTagComponent.prototype.nyActive;
    /** @type {?} */
    FarrisTagComponent.prototype.nySize;
    /** @type {?} */
    FarrisTagComponent.prototype.selectValue;
    /**
     * @type {?}
     * @private
     */
    FarrisTagComponent.prototype.el;
    /* Skipping unhandled member: ;*/
}
