// @farris/kendo-binding/esm5/lib/directives/farris-filter/farris_filter_binding.directive.js
import { Directive, Injector, Input, Optional } from '@angular/core';
import { FrameContext } from '@farris/devkit';
import { FilterComponent } from '@farris/ui-filter';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
var FarrisFilterBindingDirective = /** @class */ (function () {
    function FarrisFilterBindingDirective(injector, frameContext, filterComponent) {
        this.injector = injector;
        this.frameContext = frameContext;
        this.filterComponent = filterComponent;
        /**
         * 变量编号，默认为filterConditionList
         */
        this.variableCode = 'filterConditionList';
        this.$destroy = new Subject();
    }
    FarrisFilterBindingDirective.prototype.ngOnDestroy = function () {
        if (this.$destroy) {
            this.$destroy.next();
            this.$destroy.complete();
        }
    };
    FarrisFilterBindingDirective.prototype.ngOnInit = function () {
        var _this = this;
        if (this.filterComponent) {
            this.filterComponent.conditionsChange.pipe(takeUntil(this.$destroy)).subscribe(function (event) {
                var _a = (event || {}).data, data = _a === void 0 ? '[]' : _a;
                _this.setConditions(data);
            });
        }
    };
    FarrisFilterBindingDirective.prototype.setConditions = function (conditions) {
        this.frameContext.uiState.setPropertyValue(this.variableCode, conditions);
    };
    FarrisFilterBindingDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[farris-filter-binding]'
                },] }
    ];
    /** @nocollapse */
    FarrisFilterBindingDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: FrameContext },
        { type: FilterComponent, decorators: [{ type: Optional }] }
    ]; };
    FarrisFilterBindingDirective.propDecorators = {
        variableCode: [{ type: Input, args: ['variableCode',] }]
    };
    return FarrisFilterBindingDirective;
}());
export { FarrisFilterBindingDirective };
