// @farris/ui-datagrid/esm5/lib/components/columns/datagrid-column.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ContentChild } from '@angular/core';
import { DatagridCellEditorDirective } from './column-cell-edit.directive';
var DatagridColumnDirective = /** @class */ (function () {
    function DatagridColumnDirective() {
        this.halign = 'left';
        this.align = 'left';
        this.readonly = true;
        this.editable = false;
    }
    /**
     * @return {?}
     */
    DatagridColumnDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    DatagridColumnDirective.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
    };
    DatagridColumnDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'farris-grid-column',
                },] }
    ];
    /** @nocollapse */
    DatagridColumnDirective.ctorParameters = function () { return []; };
    DatagridColumnDirective.propDecorators = {
        field: [{ type: Input }],
        title: [{ type: Input }],
        width: [{ type: Input }],
        halign: [{ type: Input }],
        align: [{ type: Input }],
        formatter: [{ type: Input }],
        readonly: [{ type: Input }],
        editable: [{ type: Input }],
        editor: [{ type: ContentChild, args: [DatagridCellEditorDirective,] }]
    };
    return DatagridColumnDirective;
}());
export { DatagridColumnDirective };
if (false) {
    /** @type {?} */
    DatagridColumnDirective.prototype.field;
    /** @type {?} */
    DatagridColumnDirective.prototype.title;
    /** @type {?} */
    DatagridColumnDirective.prototype.width;
    /** @type {?} */
    DatagridColumnDirective.prototype.halign;
    /** @type {?} */
    DatagridColumnDirective.prototype.align;
    /** @type {?} */
    DatagridColumnDirective.prototype.formatter;
    /** @type {?} */
    DatagridColumnDirective.prototype.readonly;
    /** @type {?} */
    DatagridColumnDirective.prototype.editable;
    /** @type {?} */
    DatagridColumnDirective.prototype.editor;
}
