// @qdp/ide-cmp/esm5/lib/service/schema/schema.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { SchemaManagerService, LcpService } from '@qdp/common';
import { Subject } from 'rxjs/Subject';
var SchemaService = /** @class */ (function () {
    function SchemaService(schemaManager, lcpService) {
        var _this = this;
        this.schemaManager = schemaManager;
        this.lcpService = lcpService;
        this.schemaInfo = new Subject();
        this.colList = new Subject();
        this.pageInfoChange = new Subject();
        this.operationChange = new Subject();
        this.schemaManager.schemaInfoChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.schemaInfo.next(value);
        }));
        this.schemaManager.colList.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.colList.next(value);
        }));
        this.lcpService.pageInfoChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.pageInfoChange.next(value);
        }));
        this.schemaManager.operationChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.operationChange.next(value);
        }));
    }
    /**
     * @param {?} queryId
     * @param {?} userId
     * @param {?} organizationId
     * @param {?} queryRelativeUrl
     * @param {?=} groupType
     * @return {?}
     */
    SchemaService.prototype.getSchemaList = /**
     * @param {?} queryId
     * @param {?} userId
     * @param {?} organizationId
     * @param {?} queryRelativeUrl
     * @param {?=} groupType
     * @return {?}
     */
    function (queryId, userId, organizationId, queryRelativeUrl, groupType) {
        if (groupType) {
            return this.schemaManager.getSchemaListByGroupType(queryId, userId, organizationId, groupType, queryRelativeUrl);
        }
        else {
            return this.schemaManager.getSchemaList(queryId, userId, organizationId, queryRelativeUrl);
        }
    };
    /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @param {?=} groupType
     * @param {?=} organizationId
     * @return {?}
     */
    SchemaService.prototype.getSchema = /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @param {?=} groupType
     * @param {?=} organizationId
     * @return {?}
     */
    function (id, queryId, queryRelativeUrl, groupType, organizationId) {
        return this.schemaManager.getSchema(id, queryId, queryRelativeUrl, groupType, organizationId);
    };
    /**
     * @param {?} schema
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    SchemaService.prototype.saveSchema = /**
     * @param {?} schema
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (schema, queryRelativeUrl) {
        return this.schemaManager.saveSchema(schema, queryRelativeUrl);
    };
    /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @param {?=} groupType
     * @param {?=} organizationId
     * @return {?}
     */
    SchemaService.prototype.deleteSchema = /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @param {?=} groupType
     * @param {?=} organizationId
     * @return {?}
     */
    function (id, queryId, queryRelativeUrl, groupType, organizationId) {
        return this.schemaManager.deleteSchema(id, queryId, queryRelativeUrl, groupType, organizationId);
    };
    /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} dynamicCols
     * @param {?} queryRelativeUrl
     * @param {?=} sheme
     * @param {?=} frozenCols
     * @param {?=} groupType
     * @return {?}
     */
    SchemaService.prototype.getSpreadTemplate = /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} dynamicCols
     * @param {?} queryRelativeUrl
     * @param {?=} sheme
     * @param {?=} frozenCols
     * @param {?=} groupType
     * @return {?}
     */
    function (id, queryId, dynamicCols, queryRelativeUrl, sheme, frozenCols, groupType) {
        return this.schemaManager.getSpreadTemplate(id, queryId, dynamicCols, queryRelativeUrl, sheme, frozenCols, groupType);
    };
    /**
     * @param {?} param
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    SchemaService.prototype.getData = /**
     * @param {?} param
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (param, queryRelativeUrl) {
        return this.lcpService.getData(param, queryRelativeUrl);
    };
    /**
     * @param {?} params
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    SchemaService.prototype.getCol = /**
     * @param {?} params
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    function (params, queryRelativeUrl) {
        return this.lcpService.getCol(params, queryRelativeUrl);
    };
    /**
     * @return {?}
     */
    SchemaService.prototype.getOperation = /**
     * @return {?}
     */
    function () {
        return this.schemaManager.getOperation();
    };
    SchemaService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SchemaService.ctorParameters = function () { return [
        { type: SchemaManagerService },
        { type: LcpService }
    ]; };
    return SchemaService;
}());
export { SchemaService };
if (false) {
    /** @type {?} */
    SchemaService.prototype.schemaInfo;
    /** @type {?} */
    SchemaService.prototype.colList;
    /** @type {?} */
    SchemaService.prototype.pageInfoChange;
    /** @type {?} */
    SchemaService.prototype.operationChange;
    /**
     * @type {?}
     * @private
     */
    SchemaService.prototype.schemaManager;
    /**
     * @type {?}
     * @private
     */
    SchemaService.prototype.lcpService;
}
