// @farris/command-services/esm5/lib/entity-services/entity.service.js
import { Injectable } from '@angular/core';
import { FrameContext } from '@farris/devkit';
import { EntityTraversingService } from './entity-traversing.service';
import { EntityManipulationService } from './entity-manipulation.service';
import { EntityAggregationService } from './entity-aggregation.service';
/**
 * 实体服务
 * @scope FrameComponent
 */
var EntityService = /** @class */ (function () {
    function EntityService(frameContext) {
        this.frameContext = frameContext;
        var injector = this.frameContext.injector;
        this.traversingService = injector.get(EntityTraversingService);
        this.manipulationService = injector.get(EntityManipulationService);
        this.aggregationService = injector.get(EntityAggregationService);
    }
    /**
     * 获取path对应的属性值
     */
    EntityService.prototype.getPropValue = function (path) {
        return this.traversingService.getPropValue(path);
    };
    /**
     * 设置path对应的属性值
     */
    EntityService.prototype.setPropValue = function (path, value) {
        return this.manipulationService.setPropValue(path, value);
    };
    /**
     * 设置path对应的Entity数据
     */
    EntityService.prototype.getEntityData = function (path) {
        return this.traversingService.getEntityData(path);
    };
    /**
     * 设置path对应的EntityList数据
     */
    EntityService.prototype.getEntityListData = function (path) {
        return this.traversingService.getEntityListData(path);
    };
    /**
     * 获取path对应EntityList记录数
     */
    EntityService.prototype.count = function (path) {
        return this.aggregationService.count(path);
    };
    /**
     * 获取path对应属性的合计值
     */
    EntityService.prototype.sum = function (path) {
        return this.aggregationService.sum(path);
    };
    /**
     * 获取path对应属性的平均值
     */
    EntityService.prototype.avg = function (path) {
        return this.aggregationService.avg(path);
    };
    /**
     * 获取path对应属性的最大值
     */
    EntityService.prototype.max = function (path) {
        return this.aggregationService.max(path);
    };
    /**
     * 获取path对应属性的最小值
     */
    EntityService.prototype.min = function (path) {
        return this.aggregationService.min(path);
    };
    EntityService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    EntityService.ctorParameters = function () { return [
        { type: FrameContext }
    ]; };
    return EntityService;
}());
export { EntityService };
