// @ecp-caf/common-structure/esm5/lib/serialize-context/default/default-serialize-context.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DefaultDataTypeSerializerFactory } from './factory/default-data-type-serializer-factory';
import { DefaultOperationSerializerFactory } from './factory/default-operation-serializer-factory';
import { DefaultParameterSerializerFactory } from './factory/default-parameter-serializer-factory';
import { DefaultPropertySerializerFactory } from './factory/default-property-serializer-factory';
import { DefaultStructureSerializerFactory } from './factory/default-structure-serializer-factory';
import { DefaultConstraintSerializerFactory } from './factory/default-constraint-serializer-factory';
var DefaultSerializeContext = /** @class */ (function () {
    function DefaultSerializeContext() {
    }
    /**
     * @return {?}
     */
    DefaultSerializeContext.prototype.getDataTypeSerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.dataTypeFactory) {
            this.dataTypeFactory = new DefaultDataTypeSerializerFactory();
        }
        return this.dataTypeFactory;
    };
    /**
     * @return {?}
     */
    DefaultSerializeContext.prototype.getOperationSerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.operationFactory) {
            this.operationFactory = new DefaultOperationSerializerFactory();
        }
        return this.operationFactory;
    };
    /**
     * @return {?}
     */
    DefaultSerializeContext.prototype.getParameterSerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.parameterFactory) {
            this.parameterFactory = new DefaultParameterSerializerFactory();
        }
        return this.parameterFactory;
    };
    /**
     * @return {?}
     */
    DefaultSerializeContext.prototype.getPropertySerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.propertyFactory) {
            this.propertyFactory = new DefaultPropertySerializerFactory();
        }
        return this.propertyFactory;
    };
    /**
     * @return {?}
     */
    DefaultSerializeContext.prototype.getStructureSerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.structureFactory) {
            this.structureFactory = new DefaultStructureSerializerFactory();
        }
        return this.structureFactory;
    };
    /**
     * @return {?}
     */
    DefaultSerializeContext.prototype.getConstraintSerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.constraintFactory) {
            this.constraintFactory = new DefaultConstraintSerializerFactory();
        }
        return this.constraintFactory;
    };
    return DefaultSerializeContext;
}());
export { DefaultSerializeContext };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DefaultSerializeContext.prototype.dataTypeFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultSerializeContext.prototype.operationFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultSerializeContext.prototype.parameterFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultSerializeContext.prototype.propertyFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultSerializeContext.prototype.structureFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultSerializeContext.prototype.constraintFactory;
}
