// @farris/ide-devkit/esm5/lib/components/ng2tree-common/components/tree-node/tree-node.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Input, ComponentFactoryResolver, ViewContainerRef } from '@angular/core';
import { TreeNode } from '../../models/tree-node';
import { TreeNodeContentDirective } from '../../directives/tree-node-content.directive';
import { TreeNodeContentItem } from '../../models/tree-node-content-item';
import { TreeNodeContentComponent } from '../tree-node-content/tree-node-content.component';
var TreeNodeComponent = /** @class */ (function () {
    function TreeNodeComponent(componentFactoryResolver, viewContainerRef) {
        this.componentFactoryResolver = componentFactoryResolver;
    }
    // TODO: move to draggable directive
    // TODO: move to draggable directive
    /**
     * @param {?} $event
     * @return {?}
     */
    TreeNodeComponent.prototype.onDragStart = 
    // TODO: move to draggable directive
    /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        var _this = this;
        // 设置DragNode: 选择dragNode的父节点作为DragNode，通过index定位到指定的节点
        setTimeout(function () { return _this.node.treeModel.setDragNode({ parentNode: _this.node.parent, index: _this.nodeIndex }); }, 30);
    };
    /**
     * @return {?}
     */
    TreeNodeComponent.prototype.onDragEnd = /**
     * @return {?}
     */
    function () {
        this.node.treeModel.setDragNode(null);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    TreeNodeComponent.prototype.onDragOver = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        $event.preventDefault();
        this.node.treeModel.setDropLocation({ component: this, parentNode: this.node, index: 0 });
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    TreeNodeComponent.prototype.onDrop = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        $event.preventDefault();
        // this.node.mouseAction('drop', $event, { node: this.node, index: 0 });
        this.node.treeModel.setFocus(true);
        this.node.dropMouseAction($event, { parentNode: this.node, index: 0 });
        console.log("onDrop: tree-node-component");
    };
    /**
     * @param {?} nodeContentWrapper
     * @param {?} $event
     * @return {?}
     */
    TreeNodeComponent.prototype.onDragLeave = /**
     * @param {?} nodeContentWrapper
     * @param {?} $event
     * @return {?}
     */
    function (nodeContentWrapper, $event) {
        if (!this.node.treeModel.isDraggingOver(this)) {
            return;
        }
        /** @type {?} */
        var rect = nodeContentWrapper.getBoundingClientRect();
        // If outside the element
        if ($event.clientX < rect.left || $event.clientX > rect.right ||
            $event.clientY < rect.top || $event.clientY > rect.bottom) {
            this.node.treeModel.setDropLocation(null);
        }
    };
    /**
     * @return {?}
     */
    TreeNodeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._loadTreeNodeContent();
        // 触发变更树节点展示文本
        this.node.changeDisplayField.subscribe(function (name) {
            if (_this.componentRef) {
                _this.componentRef.instance.displayData = name;
            }
        });
    };
    /**
     * @return {?}
     */
    TreeNodeComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
    };
    // 解决使用动态组件出现ExpressionChangedAfterItHasBeenCheckedError问题：
    // https://github.com/angular/angular/issues/17572
    // 解决使用动态组件出现ExpressionChangedAfterItHasBeenCheckedError问题：
    // https://github.com/angular/angular/issues/17572
    /**
     * @return {?}
     */
    TreeNodeComponent.prototype._loadTreeNodeContent = 
    // 解决使用动态组件出现ExpressionChangedAfterItHasBeenCheckedError问题：
    // https://github.com/angular/angular/issues/17572
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        // 使用Item从逻辑上将TreeNodeContentComponent和data关联起来
        /** @type {?} */
        var treeNodeContentItem = new TreeNodeContentItem(TreeNodeContentComponent, this.node.displayField);
        // 使用 ComponentFactoryResolver 来为每个具体的组件解析出一个 ComponentFactory 
        // 然后 ComponentFactory 会为每一个组件创建一个实例
        /** @type {?} */
        var componentFactory = this.componentFactoryResolver.resolveComponentFactory(treeNodeContentItem.component);
        /** @type {?} */
        var viewContainerRef = this.treeNodeContentHost.viewContainerRef;
        viewContainerRef.clear();
        this.componentRef = viewContainerRef.createComponent(componentFactory);
        this.componentRef.instance.displayData = treeNodeContentItem.displayData; // 传入数据
        this.componentRef.instance.originData = this.node.data;
        this.componentRef.instance.rightMenuClicked.subscribe(function (event) {
            _this.node.rightMenuClicked(event);
        });
    };
    TreeNodeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng2tree-node',
                    template: "<div class=\"tree-node tree-node-level-{{ node.level }}\"\r\n[class.tree-node-expanded]=\"node.isExpanded && node.hasChildren\"\r\n[class.tree-node-collapsed]=\"node.isCollapsed && node.hasChildren\"\r\n[class.tree-node-leaf]=\"node.isLeaf\"\r\n[class.tree-node-active]=\"node.isActive\"\r\n[class.tree-node-focused]=\"node.isFocused\">\r\n\r\n<ng2tree-node-drop-slot\r\n  *ngIf=\"nodeIndex === 0\"\r\n  [dropIndex]=\"nodeIndex\"\r\n  [node]=\"node.parent\"\r\n></ng2tree-node-drop-slot>\r\n\r\n<span\r\n  *ngIf=\"node.hasChildren\"\r\n  class=\"toggle-children\"\r\n  (click)=\"node.toggleActivated()\"\r\n  (click)=\"node.toggle()\"\r\n  >\r\n</span>\r\n<span\r\n  *ngIf=\"!node.hasChildren\"\r\n  class=\"toggle-children-placeholder\">\r\n</span>\r\n\r\n<!-- \u8BBE\u7F6Eclick\u4E8B\u4EF6\u7684\u5904\u7406\u5668 -->\r\n<!-- [draggable]=\"node.allowDrag()\"\r\n(dragstart)=\"onDragStart($event)\"\r\n(drop)=\"onDrop($event)\"\r\n(dragend)=\"onDragEnd()\"\r\n(dragover)=\"onDragOver($event)\"\r\n(dragleave)=\"onDragLeave(nodeContentWrapper, $event)\" -->\r\n<div class=\"node-content-wrapper\"\r\n  #nodeContentWrapper\r\n  (click)=\"node.toggleActivated()\"\r\n  (dblclick)=\"node.doublClick($event)\"\r\n  (contextmenu)=\"node.contextMenu($event)\"\r\n\r\n  >\r\n  <!-- \u8BBE\u7F6E\u52A8\u6001\u7EC4\u4EF6\u7684\u951A\u70B9 -->\r\n  <ng-template treeNodeContent-host></ng-template> \r\n</div>\r\n\r\n<div class=\"tree-children\" [hidden]=\"node.isCollapsed\">\r\n  <ng2tree-node\r\n    *ngFor=\"let child of node.childrenField; let i = index\"\r\n    [node]=\"child\"\r\n    [nodeIndex]=\"i\">\r\n  </ng2tree-node>\r\n</div>\r\n\r\n<ng2tree-node-drop-slot\r\n  [dropIndex]=\"nodeIndex + 1\"\r\n  [node]=\"node.parent\"\r\n></ng2tree-node-drop-slot>\r\n\r\n</div>",
                    styles: [".tree-children{padding-left:20px}.node-content-wrapper{display:inline-block;padding:2px 5px;border-radius:2px;transition:background-color .15s,box-shadow .15s;white-space:nowrap}.tree-node-active>.node-content-wrapper{background:#beebff}.tree-node-active.tree-node-focused>.node-content-wrapper{background:#5db0e7}.tree-node-focused>.node-content-wrapper{background:#e7f4f9}.node-content-wrapper:hover{background:#f7fbff}.node-content-wrapper:hover,.tree-node-active>.node-content-wrapper,.tree-node-focused>.node-content-wrapper{box-shadow:inset 0 0 1px #999}.tree-node-expanded .toggle-children{background-image:url(\"data:image/svg+xml;charset=utf-8,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'%3E%3Cpath fill='%23646465' d='M11 10H5.344L11 4.414V10z'/%3E%3C/svg%3E\")}.tree-node-collapsed .toggle-children{background-image:url(\"data:image/svg+xml;charset=utf-8,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'%3E%3Cpath fill='%23646465' d='M6 4v8l4-4-4-4zm1 2.414L8.586 8 7 9.586V6.414z'/%3E%3C/svg%3E\")}.toggle-children{height:16px;width:16px;background-size:16px;display:inline-block;position:relative;background-repeat:no-repeat;background-position:center;right:-6px;top:4px;overflow:hidden}.toggle-children-placeholder{display:inline-block;height:10px;width:10px;position:relative;top:1px}.tree-node-icon{width:1rem}.tree-node{white-space:nowrap}"]
                }] }
    ];
    /** @nocollapse */
    TreeNodeComponent.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: ViewContainerRef }
    ]; };
    TreeNodeComponent.propDecorators = {
        node: [{ type: Input }],
        nodeIndex: [{ type: Input }],
        treeNodeContentHost: [{ type: ViewChild, args: [TreeNodeContentDirective,] }]
    };
    return TreeNodeComponent;
}());
export { TreeNodeComponent };
if (false) {
    /** @type {?} */
    TreeNodeComponent.prototype.node;
    /** @type {?} */
    TreeNodeComponent.prototype.nodeIndex;
    /** @type {?} */
    TreeNodeComponent.prototype.componentRef;
    /** @type {?} */
    TreeNodeComponent.prototype.treeNodeContentHost;
    /**
     * @type {?}
     * @private
     */
    TreeNodeComponent.prototype.componentFactoryResolver;
}
