// @qdp/formular/esm5/lib/component/formular-select/formular-select.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/component/formular-select/formular-select.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Output, EventEmitter } from '@angular/core';
import { MessagerService } from '@farris/ui-messager';
import { CacheService } from '@ecp-caf/caf-common';
import { RestfulService } from '../../service/restful.service';
import { EntityDataService } from '../../service/entitydata.service';
import { LocalizeService } from '@qdp/localize';
var FormularSelectComponent = /** @class */ (function () {
    function FormularSelectComponent(restfulService, cacheService, msgService, entityDataService, localizeService) {
        var _this = this;
        this.restfulService = restfulService;
        this.cacheService = cacheService;
        this.msgService = msgService;
        this.entityDataService = entityDataService;
        this.localizeService = localizeService;
        this.data = [];
        this.changeStep = new EventEmitter();
        this.formularWizard = '0'; // 公式向导类型 0公式选择 1公式新增
        // 公式向导类型 0公式选择 1公式新增
        this.ddlFormularData = []; // 下拉公示信息
        // 下拉公示信息
        this.formularDatas = []; // 所有的公式信息
        // 所有的公式信息
        this.currentFormularId = ''; // 当前选中的公式Id
        // 当前选中的公式Id
        this.currentFormularName = ''; // 当前选中的公式名称
        /**
         * 点击下一步前汇总数据
         */
        this.clickNextStep = (/**
         * @return {?}
         */
        function () {
            if (_this.formularWizard === '0' && _this.currentFormularId) {
                /** @type {?} */
                var formularData = _this.formularDatas.find((/**
                 * @param {?} x
                 * @return {?}
                 */
                function (x) { return x.id === _this.currentFormularId; }));
                if (formularData) {
                    _this.data.formularData = formularData;
                    _this.data.currentFormularName = formularData.name;
                }
                _this.data.ddlFormularData = _this.ddlFormularData;
                _this.entityDataService.setFormularState(true, { id: _this.currentFormularId, name: _this.data.currentFormularName ? _this.data.currentFormularName : '公式' });
            }
            else if (_this.formularWizard === '0' && !_this.currentFormularId) {
                _this.entityDataService.setFormularState(false, {});
                _this.msgService.info(_this.localizeService.getValue('formular.formularSelect.messageAlert.formularSelectVerify'));
                return false;
            }
            else {
                _this.entityDataService.setFormularState(false, {});
                _this.data.ddlFormularData = '';
                // this.data.formularSelectData = [];
            }
            _this.data.currentFormularId = _this.currentFormularId ? _this.currentFormularId : _this.data.currentFormularId;
            _this.data.formularWizard = _this.formularWizard;
            _this.cacheService.set('formularParam', _this.data);
            return true;
        });
        this.beforeFinish = (/**
         * @return {?}
         */
        function () {
            if (_this.formularWizard === '0' && _this.currentFormularId) {
                /** @type {?} */
                var name_1 = _this.ddlFormularData.find((/**
                 * @param {?} x
                 * @return {?}
                 */
                function (x) { return x.id === _this.currentFormularId; })).label;
                _this.entityDataService.setFormularState(true, { id: _this.currentFormularId, name: name_1 });
                // this.cacheService.set('formularResult', {id: this.currentFormularId, name: name});
                return true;
            }
            else if (_this.formularWizard === '0' && !_this.currentFormularId) {
                _this.entityDataService.setFormularState(false, {});
                _this.msgService.info(_this.localizeService.getValue('formular.formularSelect.messageAlert.formularSelectVerify'));
                return false;
            }
            else {
                _this.entityDataService.setFormularState(false, {});
                _this.msgService.info(_this.localizeService.getValue('formular.formularSelect.messageAlert.formularDefineVerify'));
                return false;
            }
        });
        this.selectionFormular = JSON.parse(this.localizeService.getValue('formular.formularSelect.selectionFormular'));
        this.lableQdpFormular = this.localizeService.getValue('formular.formularSelect.lableQdpFormular');
    }
    /**
     * @return {?}
     */
    FormularSelectComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var self = this;
        self.data = self.cacheService.get('formularParam');
        this.getFormularInfo(self);
        // if (self.data && self.data.ddlFormularData && self.data.ddlFormularData.length >0) {
        //   self.ddlFormularData = self.data.ddlFormularData;
        // } 
        if (self.data) {
            self.currentFormularId = self.data.currentFormularId ? self.data.currentFormularId : '';
            self.currentFormularName = self.data.currentFormularName ? self.data.currentFormularName : '';
            self.formularWizard = self.data.formularWizard ? self.data.formularWizard : '0';
        }
    };
    /**
     * @return {?}
     */
    FormularSelectComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @private
     * @param {?} self
     * @return {?}
     */
    FormularSelectComponent.prototype.getFormularInfo = /**
     * @private
     * @param {?} self
     * @return {?}
     */
    function (self) {
        this.createHeaderSessionId().then((/**
         * @param {?} value
         * @return {?}
         */
        function success(value) {
            /** @type {?} */
            var uri = self.data.serverHost;
            // tslint:disable-next-line: align
            /** @type {?} */
            var options = {
                headers: {
                    'SessionId': value,
                    'Content-Type': 'application/json'
                }
            };
            self.data.sessionId = value;
            /** @type {?} */
            var entityFilter = {
                FilterConditions: [{ 'FilterField': 'CurSU', 'Value': self.data.currentSU, 'Lbracket': null, 'Rbracket': null, 'Relation': 0, 'Expresstype': 0, 'Compare': 6 }],
                SortConditions: [],
                IsUsePagination: false,
                Pagination: {
                    PageIndex: 0,
                    PageSize: 0,
                    PageCount: 0,
                    TotalCount: 0
                }
            };
            uri += '/api/gsp/bcc/v1.0/querygetdatadefine_frm/extension/query?entityFilter=' + JSON.stringify(entityFilter);
            /** @type {?} */
            var ddlFormularData = [];
            self.restfulService.put(uri, {}, {}, options).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                /** @type {?} */
                var funcDatas = data.returnValue.result;
                if (funcDatas && funcDatas.length > 0) {
                    funcDatas.forEach((/**
                     * @param {?} x
                     * @return {?}
                     */
                    function (x) {
                        ddlFormularData.push({ label: x.name, id: x.id });
                    }));
                    self.ddlFormularData = tslib_1.__spread(ddlFormularData);
                }
                self.formularDatas = funcDatas;
            }));
        }));
    };
    /**
     * @private
     * @return {?}
     */
    FormularSelectComponent.prototype.createHeaderSessionId = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var options = {
            headers: {
                'Content-Type': 'application/json'
            },
            responseType: 'text'
        };
        /** @type {?} */
        var createSessionUrl = this.data.serverHost + '/api/gsp/bcc/v1.0/queryfuncdetail_frm/service/createsession';
        return this.restfulService.post(createSessionUrl, {}, {}, options).toPromise();
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    FormularSelectComponent.prototype.changeValue = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.entityDataService.setStepType($event);
        if ($event !== '0') {
            this.data.currentFormularId = '';
            this.currentFormularId = '';
            this.data.formularData = '';
            this.entityDataService.setFormularState(false, {});
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    FormularSelectComponent.prototype.selectChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if ($event) {
            this.entityDataService.setFormularState(true, { id: this.currentFormularId, name: $event.data.label });
        }
        else {
            this.entityDataService.setFormularState(false, {});
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    FormularSelectComponent.prototype.clear = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.entityDataService.setFormularState(false, {});
    };
    FormularSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-formular-select',
                    template: "<div>\r\n  <farris-radiogroup #FarrisRadioGroupComponent [(ngModel)]=\"formularWizard\" [horizontal]=\"true\"\r\n    (changeValue)=\"changeValue($event)\" [data]=\"selectionFormular\">\r\n\r\n  </farris-radiogroup>\r\n</div>\r\n<div *ngIf=\"formularWizard === '0'\">\r\n  <label class=\"col-form-label\" for=\"colFunction\" style=\"width: 6rem; min-width: 6rem;\">\r\n    <span class=\"farris-label-text\">{{lableQdpFormular}}:</span>\r\n  </label>\r\n  <div class=\"farris-input-wrap\">\r\n    <!-- <farris-combo-lookup [(ngModel)]=\"currentFormularId\" [idField]=\"'id'\" [textField]=\"'name'\"\r\n      [displayType]=\"'LOOKUPLIST'\" [multiSelect]=\"false\" [data]=\"ddlFormularData\"></farris-combo-lookup> -->\r\n    <farris-combo-list [(ngModel)]=\"currentFormularId\" (selectChange)=\"selectChange($event)\" (clear)=\"clear($event)\"\r\n      [idField]=\"'id'\" [textField]=\"'label'\" [data]=\"ddlFormularData\"></farris-combo-list>\r\n  </div>\r\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    FormularSelectComponent.ctorParameters = function () { return [
        { type: RestfulService },
        { type: CacheService },
        { type: MessagerService },
        { type: EntityDataService },
        { type: LocalizeService }
    ]; };
    FormularSelectComponent.propDecorators = {
        changeStep: [{ type: Output }]
    };
    return FormularSelectComponent;
}());
export { FormularSelectComponent };
if (false) {
    /** @type {?} */
    FormularSelectComponent.prototype.data;
    /** @type {?} */
    FormularSelectComponent.prototype.changeStep;
    /** @type {?} */
    FormularSelectComponent.prototype.formularWizard;
    /** @type {?} */
    FormularSelectComponent.prototype.ddlFormularData;
    /** @type {?} */
    FormularSelectComponent.prototype.formularDatas;
    /** @type {?} */
    FormularSelectComponent.prototype.currentFormularId;
    /** @type {?} */
    FormularSelectComponent.prototype.currentFormularName;
    /** @type {?} */
    FormularSelectComponent.prototype.selectionFormular;
    /** @type {?} */
    FormularSelectComponent.prototype.lableQdpFormular;
    /**
     * 点击下一步前汇总数据
     * @type {?}
     */
    FormularSelectComponent.prototype.clickNextStep;
    /** @type {?} */
    FormularSelectComponent.prototype.beforeFinish;
    /**
     * @type {?}
     * @private
     */
    FormularSelectComponent.prototype.restfulService;
    /**
     * @type {?}
     * @private
     */
    FormularSelectComponent.prototype.cacheService;
    /**
     * @type {?}
     * @private
     */
    FormularSelectComponent.prototype.msgService;
    /**
     * @type {?}
     * @private
     */
    FormularSelectComponent.prototype.entityDataService;
    /**
     * @type {?}
     * @private
     */
    FormularSelectComponent.prototype.localizeService;
}
