// @gsp-cmp/common-component/esm5/serialize-context/default/default-gsp-component-serialize-context.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DefaultOperationSerializerFactory, DefaultParameterSerializerFactory, DefaultPropertySerializerFactory } from '@ecp-caf/common-structure';
import { DefaultGspComponentDataTypeSerializerFactory } from './factory/default-gsp-component-data-type-serializer-factory';
var DefaultGspComponentSerializeContext = /** @class */ (function () {
    function DefaultGspComponentSerializeContext() {
    }
    // private gspComponentFactory: IGspComponentSerializerFactory;
    // private gspComponentFactory: IGspComponentSerializerFactory;
    /**
     * @return {?}
     */
    DefaultGspComponentSerializeContext.prototype.getDataTypeSerializerFactory = 
    // private gspComponentFactory: IGspComponentSerializerFactory;
    /**
     * @return {?}
     */
    function () {
        if (!this.dataTypeFactory) {
            this.dataTypeFactory = new DefaultGspComponentDataTypeSerializerFactory();
        }
        return this.dataTypeFactory;
    };
    /**
     * @return {?}
     */
    DefaultGspComponentSerializeContext.prototype.getOperationSerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.operationFactory) {
            this.operationFactory = new DefaultOperationSerializerFactory();
        }
        return this.operationFactory;
    };
    /**
     * @return {?}
     */
    DefaultGspComponentSerializeContext.prototype.getParameterSerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.parameterFactory) {
            this.parameterFactory = new DefaultParameterSerializerFactory();
        }
        return this.parameterFactory;
    };
    /**
     * @return {?}
     */
    DefaultGspComponentSerializeContext.prototype.getPropertySerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.propertyFactory) {
            this.propertyFactory = new DefaultPropertySerializerFactory();
        }
        return this.propertyFactory;
    };
    return DefaultGspComponentSerializeContext;
}());
export { DefaultGspComponentSerializeContext };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DefaultGspComponentSerializeContext.prototype.dataTypeFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultGspComponentSerializeContext.prototype.operationFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultGspComponentSerializeContext.prototype.parameterFactory;
    /**
     * @type {?}
     * @private
     */
    DefaultGspComponentSerializeContext.prototype.propertyFactory;
}
