// @farris/command-services/esm2015/lib/attachment/attachment.util.js
/**
 * 附件信息处理工具类
 */
class AttachmentUtil {
    /**
     * 转换为附件信息数组
     */
    static convertToAttachmentInfos(fileInfos) {
        if (!fileInfos) {
            return [];
        }
        const attachmentInfos = fileInfos.map((fileInfo) => {
            return this.convertToAttachmentInfo(fileInfo);
        });
        return attachmentInfos;
    }
    /**
     * 转换为附件信息
     */
    static convertToAttachmentInfo(fileInfo) {
        const attachmentInfo = {
            attachmentId: fileInfo.metadataId,
            fileName: fileInfo.fileName
        };
        return attachmentInfo;
    }
    /**
     * 获取附件列表中的第一个附件
     */
    static getFirstAttachmentInfo(attachmentInfos) {
        if (!attachmentInfos || attachmentInfos.length === 0) {
            return;
        }
        const firstAttachmentInfo = attachmentInfos[0];
        return firstAttachmentInfo;
    }
    /**
     * 提取附件id数组
     */
    static peekAttachmentIds(attachmentInfos) {
        if (!attachmentInfos) {
            attachmentInfos = [];
        }
        const attachmentIds = attachmentInfos.map((attachmentInfo) => {
            return attachmentInfo.attachmentId;
        });
        return attachmentIds;
    }
}
export { AttachmentUtil };
