// @farris/kendo-binding/esm2015/lib/directives/component-ref/combo-list-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ComboListComponent } from '@farris/ui-combo-list';
import { ComponentRef } from "./component-ref";
export class ComboListComponentRefDirective extends ComponentRef {
    constructor(injector, component, elementRef, frameContext) {
        super(frameContext, elementRef);
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngAfterViewInit() {
        this.tryAddComponent(this.component);
    }
    ngOnInit() {
        this.addComponent(this.component);
    }
    ngOnDestroy() {
    }
}
ComboListComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[comboListRef]'
            },] }
];
/** @nocollapse */
ComboListComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: ComboListComponent },
    { type: ElementRef },
    { type: FrameContext }
];
