// @farris/kendo-binding/esm2015/lib/directives/component-ref/datagrid-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { DatagridComponent } from '@farris/ui-datagrid';
import { ComponentRef } from "./component-ref";
export class DatagridComponentRefDirective extends ComponentRef {
    constructor(injector, component, elementRef, frameContext) {
        super(frameContext, elementRef);
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngAfterViewInit() {
        this.tryAddComponent(this.component, this.component.id);
    }
    ngOnInit() {
        this.addComponent(this.component);
    }
    ngOnDestroy() {
    }
}
DatagridComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[datagridRef]'
            },] }
];
/** @nocollapse */
DatagridComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: DatagridComponent },
    { type: ElementRef },
    { type: FrameContext }
];
