// @farris/kendo-binding/esm2015/lib/directives/component-ref/discussion-list-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { DiscussionListComponent } from '@farris/discussion-group';
import { ComponentRef } from "./component-ref";
export class DiscussionListComponentRefDirective extends ComponentRef {
    constructor(injector, component, elementRef, frameContext) {
        super(frameContext, elementRef);
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngAfterViewInit() {
        this.tryAddComponent(this.component);
    }
    ngOnInit() {
        this.addComponent(this.component);
    }
    ngOnDestroy() {
    }
}
DiscussionListComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[discussionListRef]'
            },] }
];
/** @nocollapse */
DiscussionListComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: DiscussionListComponent },
    { type: ElementRef },
    { type: FrameContext }
];
