// @farris/kendo-binding/esm2015/lib/directives/component-ref/discussion-editor-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { DiscussionEditorComponent } from '@farris/discussion-group';
import { ComponentRef } from "./component-ref";
export class DiscussionEditorComponentRefDirective extends ComponentRef {
    constructor(injector, component, elementRef, frameContext) {
        super(frameContext, elementRef);
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngAfterViewInit() {
        this.tryAddComponent(this.component);
    }
    ngOnInit() {
        this.addComponent(this.component);
    }
    ngOnDestroy() {
    }
}
DiscussionEditorComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[discussionEditorRef]'
            },] }
];
/** @nocollapse */
DiscussionEditorComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: DiscussionEditorComponent },
    { type: ElementRef },
    { type: FrameContext }
];
