// @farris/kendo-binding/esm2015/lib/directives/component-ref/input-group-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { InputGroupComponent } from '@farris/ui-input-group';
import { ComponentRef } from "./component-ref";
export class InputGroupComponentRefDirective extends ComponentRef {
    constructor(injector, component, elementRef, frameContext) {
        super(frameContext, elementRef);
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngAfterViewInit() {
        this.tryAddComponent(this.component);
    }
    ngOnInit() {
        this.addComponent(this.component);
    }
    ngOnDestroy() {
    }
}
InputGroupComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[inputGroupRef]'
            },] }
];
/** @nocollapse */
InputGroupComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: InputGroupComponent },
    { type: ElementRef },
    { type: FrameContext }
];
