// @farris/kendo-binding/esm2015/lib/directives/component-ref/list-nav-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ListNavComponent } from '@farris/ui-list-nav';
import { ComponentRef } from "./component-ref";
export class ListNavComponentRefDirective extends ComponentRef {
    constructor(injector, component, elementRef, frameContext) {
        super(frameContext, elementRef);
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    //#region angular lifecycle
    ngAfterViewInit() {
        this.tryAddComponent(this.component);
    }
    ngOnInit() {
        this.addComponent(this.component);
    }
    ngOnDestroy() {
    }
}
ListNavComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[listNavRef]'
            },] }
];
/** @nocollapse */
ListNavComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: ListNavComponent },
    { type: ElementRef },
    { type: FrameContext }
];
