// @farris/kendo-binding/esm2015/lib/directives/component-ref/list-view-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ListViewComponent } from "@farris/ui-list-view";
import { ComponentRef } from "./component-ref";
export class ListViewComponentRefDirective extends ComponentRef {
    constructor(injector, component, elementRef, frameContext) {
        super(frameContext, elementRef);
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngAfterViewInit() {
        this.tryAddComponent(this.component);
    }
    ngOnInit() {
        this.addComponent(this.component);
    }
    ngOnDestroy() {
    }
}
ListViewComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[listViewRef]'
            },] }
];
/** @nocollapse */
ListViewComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: ListViewComponent },
    { type: ElementRef },
    { type: FrameContext }
];
