// @farris/kendo-binding/esm2015/lib/directives/component-ref/list-filter-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ListFilterComponent } from '@farris/ui-list-filter';
import { ComponentRef } from "./component-ref";
export class ListFilterComponentRefDirective extends ComponentRef {
    constructor(injector, component, elementRef, frameContext) {
        super(frameContext, elementRef);
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngAfterViewInit() {
        this.tryAddComponent(this.component);
    }
    ngOnInit() {
        this.addComponent(this.component);
    }
    ngOnDestroy() {
    }
}
ListFilterComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[listFilterRef]'
            },] }
];
/** @nocollapse */
ListFilterComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: ListFilterComponent },
    { type: ElementRef },
    { type: FrameContext }
];
