// @farris/kendo-binding/esm2015/lib/directives/component-ref/number-spinner-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { NumberSpinnerComponent } from "@farris/ui-number-spinner";
import { ComponentRef } from "./component-ref";
export class NumberSpinnerComponentRefDirective extends ComponentRef {
    constructor(injector, component, elementRef, frameContext) {
        super(frameContext, elementRef);
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngAfterViewInit() {
        this.tryAddComponent(this.component);
    }
    ngOnInit() {
        this.addComponent(this.component);
    }
    ngOnDestroy() {
    }
}
NumberSpinnerComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[numberSpinnerRef]'
            },] }
];
/** @nocollapse */
NumberSpinnerComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: NumberSpinnerComponent },
    { type: ElementRef },
    { type: FrameContext }
];
