// @gsp-wf/task-impl-api/esm5/lib/action-base-impl/action-base-impl.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { Subject } from 'rxjs';
var ActionBaseImplComponent = /** @class */ (function () {
    function ActionBaseImplComponent() {
        /**
         * 是否为批量操作
         */
        this.isBatch = false;
    }
    /**
     * @return {?}
     */
    ActionBaseImplComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * 任务赋值
     */
    /**
     * 任务赋值
     * @param {?} t
     * @return {?}
     */
    ActionBaseImplComponent.prototype.setTask = /**
     * 任务赋值
     * @param {?} t
     * @return {?}
     */
    function (t) {
    };
    /**
     * 动作赋值
     */
    /**
     * 动作赋值
     * @param {?} a
     * @return {?}
     */
    ActionBaseImplComponent.prototype.setAction = /**
     * 动作赋值
     * @param {?} a
     * @return {?}
     */
    function (a) {
    };
    /**
     * 执行操作
     */
    /**
     * 执行操作
     * @return {?}
     */
    ActionBaseImplComponent.prototype.submit = /**
     * 执行操作
     * @return {?}
     */
    function () {
        /** @type {?} */
        var sub = new Subject();
        setTimeout(function () {
            console.log('重写动作执行逻辑');
            sub.next({
                result: true,
                isClose: true
            });
        }, 500);
        return sub.asObservable();
    };
    /**
     * 批量执行操作
     */
    /**
     * 批量执行操作
     * @param {?} tasks
     * @return {?}
     */
    ActionBaseImplComponent.prototype.batchSubmit = /**
     * 批量执行操作
     * @param {?} tasks
     * @return {?}
     */
    function (tasks) {
        /** @type {?} */
        var sub = new Subject();
        setTimeout(function () {
            console.log('重写动作执行逻辑');
            sub.next({
                result: true,
                isClose: true
            });
        }, 500);
        return sub.asObservable();
    };
    ActionBaseImplComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-action-base-impl',
                    template: "<p>\r\n  action-base-impl works!\r\n</p>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ActionBaseImplComponent.ctorParameters = function () { return []; };
    return ActionBaseImplComponent;
}());
export { ActionBaseImplComponent };
if (false) {
    /**
     * 是否为批量操作
     * @type {?}
     */
    ActionBaseImplComponent.prototype.isBatch;
    /**
     * 提交按钮名称，支持重写
     * @type {?}
     */
    ActionBaseImplComponent.prototype.submitName;
    /**
     * 任务列表上弹框的高度
     * @type {?}
     */
    ActionBaseImplComponent.prototype.dialogHeight;
    /**
     * 任务列表上弹框的宽度
     * @type {?}
     */
    ActionBaseImplComponent.prototype.dialogWidth;
    /**
     * 动作执行来源（任务列表、任务办理界面）
     * @type {?}
     */
    ActionBaseImplComponent.prototype.origin;
    /**
     * 单据
     * @type {?}
     */
    ActionBaseImplComponent.prototype.iframe;
}
