// @farris/mobile-devkit/esm5/lib/form/form_control.js
import { BindingValueAccessorFactory } from './binding_value_accessor';
import { ValidatorFactory } from '../validator/index';
/**
 * FormControl定义
 */
var FormControl = /** @class */ (function () {
    function FormControl(config, viewModelContext) {
        this.validatorFn = [];
        this.required = false;
        this.requiredExpressionFlag = false;
        this.validateExpressionFlag = false;
        this.readonly = false;
        this.visible = true;
        // 初始验证规则
        this.validRules = [];
        // 记忆验证表达式的规则
        this.validRules_validates = [];
        this.valueAccessor = BindingValueAccessorFactory.create(config.bindingType, config.bindingPath, config.valueConverter, viewModelContext);
        this.validRules = [];
        if (config.validRules) {
            this.validRules = config.validRules;
            this.setValidatorFn(config.validRules);
        }
    }
    Object.defineProperty(FormControl.prototype, "value", {
        get: function () {
            return this.valueAccessor.getValue();
        },
        set: function (val) {
            this.valueAccessor.setValue(val);
        },
        enumerable: true,
        configurable: true
    });
    FormControl.prototype.getValidRules = function () {
        return this.validRules;
    };
    FormControl.prototype.setValidatorFn = function (validRules) {
        this.validatorFn = ValidatorFactory.create(validRules);
    };
    FormControl.prototype.pushValidatorFnforValidate = function (validRules, flag) {
        if (flag && this.validateExpressionFlag) {
            return;
        }
        this.validateExpressionFlag = flag;
        this.validRules_validates = validRules;
        this.validatorFn = this.validatorFn.concat(ValidatorFactory.create(validRules));
    };
    FormControl.prototype.pushValidatorFnforRequired = function (validRules, flag) {
        if (flag && this.requiredExpressionFlag) {
            return;
        }
        this.requiredExpressionFlag = flag;
        this.validatorFn = this.validatorFn.concat(ValidatorFactory.create(validRules));
    };
    FormControl.prototype.resetValidatorFnforRequired = function () {
        this.requiredExpressionFlag = false;
        this.validatorFn = ValidatorFactory.create(this.validRules.concat(this.validRules_validates));
    };
    FormControl.prototype.getValidatorFn = function () {
        return this.validatorFn;
    };
    return FormControl;
}());
export { FormControl };
