// @farris/dynamic-control-group/esm5/lib/components/number/number-range.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from "@angular/core";
import { FormGroup } from "@angular/forms";
var NumberRangeComponent = /** @class */ (function () {
    function NumberRangeComponent() {
    }
    /**
     * @return {?}
     */
    NumberRangeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.options = this.control.control;
        this.className = this.control.className ? this.control.className : 'col-12 col-md-6 col-xl-3 col-el-2';
    };
    NumberRangeComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-number-range",
                    template: "\n    <div class=\"farris-group-wrap\" [formGroup]=\"form\">\n      <div class=\"form-group farris-form-group\">\n        <label class=\"col-form-label\">\n          <span *ngIf=\"0\" class=\"farris-label-info text-danger\">*</span>\n          <span class=\"farris-label-text\">{{control.name}}</span>\n        </label>\n        <div class=\"farris-input-wrap\">\n            <farris-number-range\n                [formControlName]=\"control.labelCode\"\n                [max]=\"options.max\"\n                [min]=\"options.min\"\n                [precision]=\"options.precision\"\n                [readonly]=\"options.readonly\"\n            >\n            </farris-number-range>\n        </div>\n      </div>\n    </div>\n  "
                }] }
    ];
    /** @nocollapse */
    NumberRangeComponent.ctorParameters = function () { return []; };
    NumberRangeComponent.propDecorators = {
        className: [{ type: HostBinding, args: ['class',] }],
        control: [{ type: Input }],
        form: [{ type: Input }]
    };
    return NumberRangeComponent;
}());
export { NumberRangeComponent };
if (false) {
    /** @type {?} */
    NumberRangeComponent.prototype.className;
    /** @type {?} */
    NumberRangeComponent.prototype.control;
    /** @type {?} */
    NumberRangeComponent.prototype.form;
    /** @type {?} */
    NumberRangeComponent.prototype.options;
}
