// @edp-pmf/mxgraph-ts/dist/cjs/lib/Handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mxVertexHandler = exports.mxTooltipHandler = exports.mxSelectionCellsHandler = exports.mxRubberband = exports.mxPopupMenuHandler = exports.mxPanningHandler = exports.mxKeyHandler = exports.mxHandle = exports.mxGraphHandler = exports.mxElbowEdgeHandler = exports.mxEdgeSegmentHandler = exports.mxEdgeHandler = exports.mxConstraintHandler = exports.mxConnectionHandler = exports.mxCellTracker = exports.mxCellMarker = exports.mxCellHighlight = void 0;
var mxgraphFactory_1 = require("./mxgraphFactory");
/**
 * Handler
 */
exports.mxCellHighlight = mxgraphFactory_1.mxgraphInstance.mxCellHighlight;
exports.mxCellMarker = mxgraphFactory_1.mxgraphInstance.mxCellMarker;
exports.mxCellTracker = mxgraphFactory_1.mxgraphInstance.mxCellTracker;
exports.mxConnectionHandler = mxgraphFactory_1.mxgraphInstance.mxConnectionHandler;
exports.mxConstraintHandler = mxgraphFactory_1.mxgraphInstance.mxConstraintHandler;
exports.mxEdgeHandler = mxgraphFactory_1.mxgraphInstance.mxEdgeHandler;
exports.mxEdgeSegmentHandler = mxgraphFactory_1.mxgraphInstance.mxEdgeSegmentHandler;
exports.mxElbowEdgeHandler = mxgraphFactory_1.mxgraphInstance.mxElbowEdgeHandler;
exports.mxGraphHandler = mxgraphFactory_1.mxgraphInstance.mxGraphHandler;
exports.mxHandle = mxgraphFactory_1.mxgraphInstance.mxHandle;
exports.mxKeyHandler = mxgraphFactory_1.mxgraphInstance.mxKeyHandler;
exports.mxPanningHandler = mxgraphFactory_1.mxgraphInstance.mxPanningHandler;
exports.mxPopupMenuHandler = mxgraphFactory_1.mxgraphInstance.mxPopupMenuHandler;
exports.mxRubberband = mxgraphFactory_1.mxgraphInstance.mxRubberband;
exports.mxSelectionCellsHandler = mxgraphFactory_1.mxgraphInstance.mxSelectionCellsHandler;
exports.mxTooltipHandler = mxgraphFactory_1.mxgraphInstance.mxTooltipHandler;
exports.mxVertexHandler = mxgraphFactory_1.mxgraphInstance.mxVertexHandler;
