// @edp-pmf/mxgraph-ts/dist/cjs/lib/IO.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mxStylesheetCodec = exports.mxObjectCodec = exports.mxGenericChangeCodec = exports.mxDefaultToolbarCodec = exports.mxCodecRegistry = exports.mxCodec = void 0;
var mxgraphFactory_1 = require("./mxgraphFactory");
/**
 * IO
 */
// export const mxCellCodec = mxgraphInstance.mxCellCodec;
// export const mxChildChangeCodec = mxgraphInstance.mxChildChangeCodec;
exports.mxCodec = mxgraphFactory_1.mxgraphInstance.mxCodec;
exports.mxCodecRegistry = mxgraphFactory_1.mxgraphInstance.mxCodecRegistry;
// export const mxDefaultKeyHandlerCodec = mxgraphInstance.mxDefaultKeyHandlerCodec;
// export const mxDefaultPopupMenuCodec = mxgraphInstance.mxDefaultPopupMenuCodec;
exports.mxDefaultToolbarCodec = mxgraphFactory_1.mxgraphInstance.mxDefaultToolbarCodec;
// export const mxEditorCodec = mxgraphInstance.mxEditorCodec;
exports.mxGenericChangeCodec = mxgraphFactory_1.mxgraphInstance.mxGenericChangeCodec;
// export const mxGraphCodec = mxgraphInstance.mxGraphCodec;
// export const mxGraphViewCodec = mxgraphInstance.mxGraphViewCodec;
// export const mxModelCodec = mxgraphInstance.mxModelCodec;
exports.mxObjectCodec = mxgraphFactory_1.mxgraphInstance.mxObjectCodec;
// export const mxRootChangeCodec = mxgraphInstance.mxRootChangeCodec;
exports.mxStylesheetCodec = mxgraphFactory_1.mxgraphInstance.mxStylesheetCodec;
// export const mxTerminalChangeCodec = mxgraphInstance.mxTerminalChangeCodec;
