// @edp-pmf/mxgraph-ts/dist/cjs/lib/Layout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mxSwimlaneOrdering = exports.mxMinimumCycleRemover = exports.mxMedianHybridCrossingReduction = exports.mxHierarchicalLayoutStage = exports.mxCoordinateAssignment = exports.mxSwimlaneModel = exports.mxGraphHierarchyNode = exports.mxGraphHierarchyModel = exports.mxGraphHierarchyEdge = exports.mxGraphAbstractHierarchyCell = exports.mxSwimlaneLayout = exports.mxHierarchicalLayout = exports.mxStackLayout = exports.mxRadialTreeLayout = exports.mxPartitionLayout = exports.mxParallelEdgeLayout = exports.mxGraphLayout = exports.mxFastOrganicLayout = exports.mxEdgeLabelLayout = exports.mxCompositeLayout = exports.mxCompactTreeLayout = exports.mxCircleLayout = void 0;
var mxgraphFactory_1 = require("./mxgraphFactory");
/**
 * Layout
 */
exports.mxCircleLayout = mxgraphFactory_1.mxgraphInstance.mxCircleLayout;
exports.mxCompactTreeLayout = mxgraphFactory_1.mxgraphInstance.mxCompactTreeLayout;
exports.mxCompositeLayout = mxgraphFactory_1.mxgraphInstance.mxCompositeLayout;
exports.mxEdgeLabelLayout = mxgraphFactory_1.mxgraphInstance.mxEdgeLabelLayout;
exports.mxFastOrganicLayout = mxgraphFactory_1.mxgraphInstance.mxFastOrganicLayout;
exports.mxGraphLayout = mxgraphFactory_1.mxgraphInstance.mxGraphLayout;
exports.mxParallelEdgeLayout = mxgraphFactory_1.mxgraphInstance.mxParallelEdgeLayout;
exports.mxPartitionLayout = mxgraphFactory_1.mxgraphInstance.mxPartitionLayout;
exports.mxRadialTreeLayout = mxgraphFactory_1.mxgraphInstance.mxRadialTreeLayout;
exports.mxStackLayout = mxgraphFactory_1.mxgraphInstance.mxStackLayout;
/**
* Layout Hierarchical
*/
exports.mxHierarchicalLayout = mxgraphFactory_1.mxgraphInstance.mxHierarchicalLayout;
exports.mxSwimlaneLayout = mxgraphFactory_1.mxgraphInstance.mxSwimlaneLayout;
/**
* Layout Hierarchical Model
*/
exports.mxGraphAbstractHierarchyCell = mxgraphFactory_1.mxgraphInstance.mxGraphAbstractHierarchyCell;
exports.mxGraphHierarchyEdge = mxgraphFactory_1.mxgraphInstance.mxGraphHierarchyEdge;
exports.mxGraphHierarchyModel = mxgraphFactory_1.mxgraphInstance.mxGraphHierarchyModel;
exports.mxGraphHierarchyNode = mxgraphFactory_1.mxgraphInstance.mxGraphHierarchyNode;
exports.mxSwimlaneModel = mxgraphFactory_1.mxgraphInstance.mxSwimlaneModel;
/**
* Layout Hierarchical Stage
*/
exports.mxCoordinateAssignment = mxgraphFactory_1.mxgraphInstance.mxCoordinateAssignment;
exports.mxHierarchicalLayoutStage = mxgraphFactory_1.mxgraphInstance.mxHierarchicalLayoutStage;
exports.mxMedianHybridCrossingReduction = mxgraphFactory_1.mxgraphInstance.mxMedianHybridCrossingReduction;
exports.mxMinimumCycleRemover = mxgraphFactory_1.mxgraphInstance.mxMinimumCycleRemover;
exports.mxSwimlaneOrdering = mxgraphFactory_1.mxgraphInstance.mxSwimlaneOrdering;
