// @edp-pmf/mxgraph-ts/dist/cjs/lib/Model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mxGraphModel = exports.mxGeometry = exports.mxCellPath = exports.mxCell = void 0;
var mxgraphFactory_1 = require("./mxgraphFactory");
/**
 * Model
 */
exports.mxCell = mxgraphFactory_1.mxgraphInstance.mxCell;
exports.mxCellPath = mxgraphFactory_1.mxgraphInstance.mxCellPath;
exports.mxGeometry = mxgraphFactory_1.mxgraphInstance.mxGeometry;
exports.mxGraphModel = mxgraphFactory_1.mxgraphInstance.mxGraphModel;
// /**
//  * mxGraphModel Change
//  */
// export const mxRootChange = mxgraphInstance.mxRootChange;
// export const mxChildChange = mxgraphInstance.mxChildChange;
// export const mxTerminalChange = mxgraphInstance.mxTerminalChange;
// export const mxValueChange = mxgraphInstance.mxValueChange;
// export const mxStyleChange = mxgraphInstance.mxStyleChange;
// export const mxGeometryChange = mxgraphInstance.mxGeometryChange;
// export const mxCollapseChange = mxgraphInstance.mxCollapseChange;
// export const mxVisibleChange = mxgraphInstance.mxVisibleChange;
// export const mxCellAttributeChange = mxgraphInstance.mxCellAttributeChange;
