// @edp-pmf/mxgraph-ts/dist/cjs/lib/Shape.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mxTriangle = exports.mxText = exports.mxSwimlane = exports.mxStencilRegistry = exports.mxStencil = exports.mxShape = exports.mxRhombus = exports.mxRectangleShape = exports.mxPolyline = exports.mxMarker = exports.mxLine = exports.mxLabel = exports.mxImageShape = exports.mxHexagon = exports.mxEllipse = exports.mxDoubleEllipse = exports.mxCylinder = exports.mxConnector = exports.mxCloud = exports.mxArrowConnector = exports.mxArrow = exports.mxActor = void 0;
var mxgraphFactory_1 = require("./mxgraphFactory");
/**
 * Shape
 */
exports.mxActor = mxgraphFactory_1.mxgraphInstance.mxActor;
exports.mxArrow = mxgraphFactory_1.mxgraphInstance.mxArrow;
exports.mxArrowConnector = mxgraphFactory_1.mxgraphInstance.mxArrowConnector;
exports.mxCloud = mxgraphFactory_1.mxgraphInstance.mxCloud;
exports.mxConnector = mxgraphFactory_1.mxgraphInstance.mxConnector;
exports.mxCylinder = mxgraphFactory_1.mxgraphInstance.mxCylinder;
exports.mxDoubleEllipse = mxgraphFactory_1.mxgraphInstance.mxDoubleEllipse;
exports.mxEllipse = mxgraphFactory_1.mxgraphInstance.mxEllipse;
exports.mxHexagon = mxgraphFactory_1.mxgraphInstance.mxHexagon;
exports.mxImageShape = mxgraphFactory_1.mxgraphInstance.mxImageShape;
exports.mxLabel = mxgraphFactory_1.mxgraphInstance.mxLabel;
exports.mxLine = mxgraphFactory_1.mxgraphInstance.mxLine;
exports.mxMarker = mxgraphFactory_1.mxgraphInstance.mxMarker;
exports.mxPolyline = mxgraphFactory_1.mxgraphInstance.mxPolyline;
exports.mxRectangleShape = mxgraphFactory_1.mxgraphInstance.mxRectangleShape;
exports.mxRhombus = mxgraphFactory_1.mxgraphInstance.mxRhombus;
exports.mxShape = mxgraphFactory_1.mxgraphInstance.mxShape;
exports.mxStencil = mxgraphFactory_1.mxgraphInstance.mxStencil;
exports.mxStencilRegistry = mxgraphFactory_1.mxgraphInstance.mxStencilRegistry;
exports.mxSwimlane = mxgraphFactory_1.mxgraphInstance.mxSwimlane;
exports.mxText = mxgraphFactory_1.mxgraphInstance.mxText;
exports.mxTriangle = mxgraphFactory_1.mxgraphInstance.mxTriangle;
