// @edp-pmf/mxgraph-ts/dist/cjs/lib/Util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mxXmlRequest = exports.mxXmlCanvas2D = exports.mxWindow = exports.mxVmlCanvas2D = exports.mxUtils = exports.mxUrlConverter = exports.mxUndoManager = exports.mxUndoableEdit = exports.mxToolbar = exports.mxSvgCanvas2D = exports.mxResources = exports.mxRectangle = exports.mxPopupMenu = exports.mxPoint = exports.mxPanningManager = exports.mxObjectIdentity = exports.mxMouseEvent = exports.mxMorphing = exports.mxLog = exports.mxImageExport = exports.mxImageBundle = exports.mxImage = exports.mxGuide = exports.mxForm = exports.mxEventSource = exports.mxEventObject = exports.mxEvent = exports.mxEffects = exports.mxDragSource = exports.mxDivResizer = exports.mxDictionary = exports.mxConstants = exports.mxClipboard = exports.mxAutoSaveManager = exports.mxAnimation = exports.mxAbstractCanvas2D = void 0;
var mxgraphFactory_1 = require("./mxgraphFactory");
/**
 * Util
 */
exports.mxAbstractCanvas2D = mxgraphFactory_1.mxgraphInstance.mxAbstractCanvas2D;
exports.mxAnimation = mxgraphFactory_1.mxgraphInstance.mxAnimation;
exports.mxAutoSaveManager = mxgraphFactory_1.mxgraphInstance.mxAutoSaveManager;
exports.mxClipboard = mxgraphFactory_1.mxgraphInstance.mxClipboard;
exports.mxConstants = mxgraphFactory_1.mxgraphInstance.mxConstants;
exports.mxDictionary = mxgraphFactory_1.mxgraphInstance.mxDictionary;
exports.mxDivResizer = mxgraphFactory_1.mxgraphInstance.mxDivResizer;
exports.mxDragSource = mxgraphFactory_1.mxgraphInstance.mxDragSource;
exports.mxEffects = mxgraphFactory_1.mxgraphInstance.mxEffects;
exports.mxEvent = mxgraphFactory_1.mxgraphInstance.mxEvent;
exports.mxEventObject = mxgraphFactory_1.mxgraphInstance.mxEventObject;
exports.mxEventSource = mxgraphFactory_1.mxgraphInstance.mxEventSource;
exports.mxForm = mxgraphFactory_1.mxgraphInstance.mxForm;
exports.mxGuide = mxgraphFactory_1.mxgraphInstance.mxGuide;
exports.mxImage = mxgraphFactory_1.mxgraphInstance.mxImage;
exports.mxImageBundle = mxgraphFactory_1.mxgraphInstance.mxImageBundle;
exports.mxImageExport = mxgraphFactory_1.mxgraphInstance.mxImageExport;
exports.mxLog = mxgraphFactory_1.mxgraphInstance.mxLog;
exports.mxMorphing = mxgraphFactory_1.mxgraphInstance.mxMorphing;
exports.mxMouseEvent = mxgraphFactory_1.mxgraphInstance.mxMouseEvent;
exports.mxObjectIdentity = mxgraphFactory_1.mxgraphInstance.mxObjectIdentity;
exports.mxPanningManager = mxgraphFactory_1.mxgraphInstance.mxPanningManager;
exports.mxPoint = mxgraphFactory_1.mxgraphInstance.mxPoint;
exports.mxPopupMenu = mxgraphFactory_1.mxgraphInstance.mxPopupMenu;
exports.mxRectangle = mxgraphFactory_1.mxgraphInstance.mxRectangle;
exports.mxResources = mxgraphFactory_1.mxgraphInstance.mxResources;
exports.mxSvgCanvas2D = mxgraphFactory_1.mxgraphInstance.mxSvgCanvas2D;
exports.mxToolbar = mxgraphFactory_1.mxgraphInstance.mxToolbar;
exports.mxUndoableEdit = mxgraphFactory_1.mxgraphInstance.mxUndoableEdit;
exports.mxUndoManager = mxgraphFactory_1.mxgraphInstance.mxUndoManager;
exports.mxUrlConverter = mxgraphFactory_1.mxgraphInstance.mxUrlConverter;
exports.mxUtils = mxgraphFactory_1.mxgraphInstance.mxUtils;
exports.mxVmlCanvas2D = mxgraphFactory_1.mxgraphInstance.mxVmlCanvas2D;
exports.mxWindow = mxgraphFactory_1.mxgraphInstance.mxWindow;
exports.mxXmlCanvas2D = mxgraphFactory_1.mxgraphInstance.mxXmlCanvas2D;
exports.mxXmlRequest = mxgraphFactory_1.mxgraphInstance.mxXmlRequest;
