// @edp-pmf/mxgraph-ts/dist/cjs/lib/View.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mxTemporaryCellStates = exports.mxSwimlaneManager = exports.mxStylesheet = exports.mxStyleRegistry = exports.mxPrintPreview = exports.mxPerimeter = exports.mxOutline = exports.mxMultiplicity = exports.mxLayoutManager = exports.mxGraphView = exports.mxGraphSelectionModel = exports.mxGraph = exports.mxEdgeStyle = exports.mxConnectionConstraint = exports.mxCellStatePreview = exports.mxCellState = exports.mxCellRenderer = exports.mxCellOverlay = exports.mxCellEditor = void 0;
var mxgraphFactory_1 = require("./mxgraphFactory");
/**
 * View
 */
exports.mxCellEditor = mxgraphFactory_1.mxgraphInstance.mxCellEditor;
exports.mxCellOverlay = mxgraphFactory_1.mxgraphInstance.mxCellOverlay;
exports.mxCellRenderer = mxgraphFactory_1.mxgraphInstance.mxCellRenderer;
exports.mxCellState = mxgraphFactory_1.mxgraphInstance.mxCellState;
exports.mxCellStatePreview = mxgraphFactory_1.mxgraphInstance.mxCellStatePreview;
exports.mxConnectionConstraint = mxgraphFactory_1.mxgraphInstance.mxConnectionConstraint;
exports.mxEdgeStyle = mxgraphFactory_1.mxgraphInstance.mxEdgeStyle;
exports.mxGraph = mxgraphFactory_1.mxgraphInstance.mxGraph;
exports.mxGraphSelectionModel = mxgraphFactory_1.mxgraphInstance.mxGraphSelectionModel;
exports.mxGraphView = mxgraphFactory_1.mxgraphInstance.mxGraphView;
exports.mxLayoutManager = mxgraphFactory_1.mxgraphInstance.mxLayoutManager;
exports.mxMultiplicity = mxgraphFactory_1.mxgraphInstance.mxMultiplicity;
exports.mxOutline = mxgraphFactory_1.mxgraphInstance.mxOutline;
exports.mxPerimeter = mxgraphFactory_1.mxgraphInstance.mxPerimeter;
exports.mxPrintPreview = mxgraphFactory_1.mxgraphInstance.mxPrintPreview;
exports.mxStyleRegistry = mxgraphFactory_1.mxgraphInstance.mxStyleRegistry;
exports.mxStylesheet = mxgraphFactory_1.mxgraphInstance.mxStylesheet;
exports.mxSwimlaneManager = mxgraphFactory_1.mxgraphInstance.mxSwimlaneManager;
exports.mxTemporaryCellStates = mxgraphFactory_1.mxgraphInstance.mxTemporaryCellStates;
// export const renderingHint = mxgraphInstance.renderingHint;
/**
 * mxGraphSelectionModel
 */
// export const mxSelectionChange = mxgraphInstance.mxSelectionChange;
/**
 * mxGraphView
 */
// export const mxCurrentRootChange = mxgraphInstance.mxCurrentRootChange;
