// @farris/ui-button/esm2015/lib/button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output, ViewChild, ElementRef, Optional, Injector } from '@angular/core';
export class FarrisButtonComponent {
    /**
     * @param {?} injector
     * @param {?} _elRef
     */
    constructor(injector, _elRef) {
        this.injector = injector;
        this._elRef = _elRef;
        this.size = 'small';
        this.type = 'primary';
        this.click = new EventEmitter();
        this.__zone_click = null;
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.el = _elRef ? _elRef.nativeElement : this.btnEl ? this.btnEl.nativeElement.parentElement : null;
            this.initButtonEvent();
        }));
    }
    /**
     * @private
     * @return {?}
     */
    initButtonEvent() {
        if (this.el && this.el.closest('grid-body-cell')) {
            this.el.addEventListener('mousedown', this.onBtnMouseDown.bind(this), true);
        }
    }
    /**
     * @private
     * @param {?} e
     * @return {?}
     */
    onBtnMouseDown(e) {
        this.__zone_click = this.el['__zone_symbol__clickfalse'];
        if (this.__zone_click && this.__zone_click.length) {
            e.currentTarget.style.pointerEvents = 'none';
            this.el['__zone_symbol__clickfalse'] = null;
        }
    }
    /**
     * @param {?} e
     * @return {?}
     */
    clickEvent(e) {
        e.stopPropagation();
        if (this.disabled) {
            return;
        }
        this.click.emit(e);
    }
}
FarrisButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'farris-button',
                template: `
    <button [class]="(size === 'large'?'btn-lg ':'btn-sm ')+
    'btn '+ (type?'btn-'+type+' ':'')"  #btn 
    [class.disabled]="disabled" (click)="clickEvent($event)">
      <ng-content></ng-content>
    </button>
  `
            }] }
];
/** @nocollapse */
FarrisButtonComponent.ctorParameters = () => [
    { type: Injector, decorators: [{ type: Optional }] },
    { type: ElementRef, decorators: [{ type: Optional }] }
];
FarrisButtonComponent.propDecorators = {
    text: [{ type: Input }],
    size: [{ type: Input }],
    type: [{ type: Input }],
    disabled: [{ type: Input }],
    click: [{ type: Output }],
    btnEl: [{ type: ViewChild, args: ['btn',] }]
};
if (false) {
    /** @type {?} */
    FarrisButtonComponent.prototype.text;
    /** @type {?} */
    FarrisButtonComponent.prototype.size;
    /** @type {?} */
    FarrisButtonComponent.prototype.type;
    /** @type {?} */
    FarrisButtonComponent.prototype.disabled;
    /** @type {?} */
    FarrisButtonComponent.prototype.click;
    /** @type {?} */
    FarrisButtonComponent.prototype.btnEl;
    /** @type {?} */
    FarrisButtonComponent.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FarrisButtonComponent.prototype.__zone_click;
    /**
     * @type {?}
     * @private
     */
    FarrisButtonComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    FarrisButtonComponent.prototype._elRef;
}
