// @gsp-svc/file-viewer/esm2015/lib/gsp-viewer/gsp-viewer.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { GspViewerComponent } from './gsp-viewer.component';
import { ImageViewerModule } from '../image-viewer/image-viewer.module';
import { DocViewerModule } from '../doc-viewer/doc-viewer.module';
import { PdfViewerModule } from '../pdf-viewer/pdf-viewer.module';
import { OfdViewerModule } from '../ofd-viewer/ofd-viewer.module';
import { NotifyModule } from '@farris/ui-notify';
import { FileListService } from '../service/file-list.service';
import { RotationDirective } from '../service/rotation.directive';
import { SearchService } from '../service/search.service';
import { SelectionService } from '../service/selection.service';
import { ZoomService } from '../service/zoom.service';
import { ConfigService } from '../service/config.service';
import { DocViewerService } from '../service/doc-viewer.service';
import { NavigateService } from '../service/navigate.service';
import { PagePreloadService } from '../service/page-preload.service';
import { WindowService } from '../service/window.service';
export class GspViewerModule {
}
GspViewerModule.decorators = [
    { type: NgModule, args: [{
                declarations: [GspViewerComponent],
                imports: [
                    CommonModule,
                    ImageViewerModule,
                    DocViewerModule,
                    OfdViewerModule,
                    PdfViewerModule,
                    NotifyModule.forRoot()
                ],
                exports: [GspViewerComponent],
                entryComponents: [GspViewerComponent],
                providers: [RotationDirective, SearchService, SelectionService, ZoomService, ConfigService, DocViewerService,
                    FileListService, NavigateService, PagePreloadService, WindowService]
            },] }
];
