// @farris/bef/esm2015/lib/bef_data_path_util.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_data_path_util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BindingPropertyType, PropertyUtil, BindingPathConverter } from '@farris/devkit';
/**
 * Bef数据中各种Path的转换
 */
class BefDataPathUtil {
    /**
     * 转换成路径数组
     * 返回结果：
     * 主表（/）：[ parentId ]
     * 从表（/childCodes）：[ parentId, childCodes, childId ]
     * 从从表（/childCodes/grandCodes）： [ parentId, childCode, childId, grandCode, grandId ]
     * @param {?} bindingPath
     * @param {?} bindingData
     * @return {?}
     */
    static convertToPathArray(bindingPath, bindingData) {
        /** @type {?} */
        const bindingPathArray = BindingPathConverter.toBindingPathArray(bindingPath);
        /** @type {?} */
        const pathArray = [];
        /** @type {?} */
        let currentBindingObject = bindingData.list.currentItem;
        pathArray.push(currentBindingObject.primaryKeyValue);
        bindingPathArray.forEach((/**
         * @param {?} propName
         * @return {?}
         */
        (propName) => {
            /** @type {?} */
            const propInfo = PropertyUtil.getPropertyByName(currentBindingObject.properties, propName);
            if (propInfo.type !== BindingPropertyType.List) {
                throw new Error(`${propInfo.name}不是子表对应的属性`);
            }
            /** @type {?} */
            const currentBindingList = currentBindingObject[propName];
            currentBindingObject = currentBindingList.currentItem;
            pathArray.push(this.trimTrailingS(propName));
            pathArray.push(currentBindingObject.primaryKeyValue);
        }));
        return pathArray;
    }
    /**
     * 转换为RestUrl里的路径
     *
     * 返回结果：
     * 主表（/）：/
     * 从表（/jiwtEdus）：/xxx/jiwtEdu
     * 从从表（/jiwtEdus/jiwtGrades）： /xxx/jiwtEdu/xxx/jiwtGrade
     * @param {?} bindingPath
     * @param {?} bindingData
     * @return {?}
     */
    static convertToPathUrl(bindingPath, bindingData) {
        /** @type {?} */
        const pathArray = this.convertToPathArray(bindingPath, bindingData);
        pathArray.pop();
        return '/' + pathArray.join('/');
    }
    /**
     * 转换为表名数组
     * 返回结果：
     * 主表（/）：[]
     * 从表（/childCodes）：[ childCodes ]
     * 从从表（/childCodes/grandCodes [childCode ,grandCode]
     * @param {?} bindingPath
     * @param {?} bindingData
     * @return {?}
     */
    static convertToObjectCodes(bindingPath, bindingData) {
        /** @type {?} */
        const pathArray = this.convertToPathArray(bindingPath, bindingData);
        /** @type {?} */
        const pathLen = pathArray.length;
        /** @type {?} */
        const objectCodes = [];
        for (let i = 1; i < pathLen; i = i + 2) {
            objectCodes.push(pathArray[i]);
        }
        return objectCodes;
    }
    /**
     * 转换为id数组，包含最后一级的主键
     * 主表（/）：[ parentId ]
     * 从表（/childCodes）：[ parentId, childId ]
     * 从从表（/childCodes/grandCodes）： [ parentId, childId, grandId]
     * @param {?} bindingPath
     * @param {?} bindingData
     * @return {?}
     */
    static convertToDataIdsForUpdate(bindingPath, bindingData) {
        /** @type {?} */
        const pathArray = this.convertToPathArray(bindingPath, bindingData);
        /** @type {?} */
        const pathLen = pathArray.length;
        /** @type {?} */
        const dataIds = [];
        for (let i = 0; i < pathLen; i = i + 2) {
            dataIds.push(pathArray[i]);
        }
        return dataIds;
    }
    /**
     * 转换为id数组，不包含最后一级的主键
     * 主表（/）：[ ]
     * 从表（/childCodes）：[ parentId ]
     * 从从表（/childCodes/grandCodes）： [ parentId, childId]
     * @param {?} bindingPath
     * @param {?} bindingData
     * @return {?}
     */
    static convertToDataIdsForAdd(bindingPath, bindingData) {
        /** @type {?} */
        const dataIds = this.convertToDataIdsForUpdate(bindingPath, bindingData);
        dataIds.pop();
        return dataIds;
    }
    /**
     * 将EntityPathString转换为Url
     * @deprecated
     * @param {?} path
     * @return {?}
     */
    static convertPathToUrl(path) {
        /** @type {?} */
        const subPaths = path.split('/');
        for (let i = subPaths.length - 1; i > 0; i--) {
            if (subPaths[i] && subPaths[i].endsWith('s')) {
                subPaths[i] = subPaths[i].substr(0, subPaths[i].length - 1).toLowerCase();
            }
        }
        // url不再转小写，兼容主键有大写字母的情况
        return subPaths.join('/'); // .toLowerCase();
    }
    /**
     * 去除最后的s
     * @private
     * @param {?} str
     * @return {?}
     */
    static trimTrailingS(str) {
        return str.substr(0, str.length - 1);
    }
    /**
     * 转换通用路径为仅有表名的数组
     *
     * 1.['id:xxx','child1s','id:xx','prop'] => ['child1s','prop']
     * 2.['prop1','prop2','prop3'] => ['prop1','prop2','prop3']
     * 3.['prop1','id:xxx','prop2'] => ['prop1','prop2']
     * @param {?} paths 内置路径，格式:['id:xxx','child1s','id:xx','prop']
     * @return {?}
     */
    static convertPathsToNodeCodes(paths) {
        /** @type {?} */
        const result = [];
        if (!paths || paths.length < 1) {
            return result;
        }
        paths.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => !!item && item.indexOf(':') === -1)).forEach((/**
         * @param {?} node
         * @return {?}
         */
        node => {
            result.push(node);
        }));
        return result;
    }
    /**
     * 转换通用路径为仅有实体主键值的数组
     * @param {?} paths 通用路径，格式:['id:xxx','child1s','id:xx','prop'] => ['xxx','xx']
     * @return {?}
     */
    static convertPathsToIds(paths) {
        /** @type {?} */
        const result = [];
        if (!paths || paths.length < 1) {
            return result;
        }
        paths.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => !!item && item.indexOf(':') !== -1)).forEach((/**
         * @param {?} id
         * @return {?}
         */
        id => {
            result.push(id.split(':')[1]);
        }));
        return result;
    }
}
export { BefDataPathUtil };
