// @farris/bef/esm2015/lib/bef_entity_manager.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bef_entity_manager.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: Witt
 * @Date: 2019-03-07 17:24:38
 * @Last Modified by:   Witt
 * @Last Modified time: 2019-03-07 17:24:38
 */
import { EntityManager } from '@farris/devkit';
import { BefChangeBuilder } from './bef_change_builder';
import { BefChangeHandler } from './bef_change_handler';
/**
 * 实体管理类：管理Entity和
 * \@todo
 * 1、实体管理应该放在Devkit中；
 * 2、目前只是简单代码拆分，待进一步明确职责。
 * @template T
 */
class BefEntityManager extends EntityManager {
    /**
     * 构造函数
     * @param {?} entityCollection
     */
    constructor(entityCollection) {
        super(entityCollection);
        this.changeBuilder = new BefChangeBuilder(this.entityCollection.entityType, this.entityCollection);
        this.changeHandler = new BefChangeHandler();
    }
    // #region 处理ChangeDetail相关
    /**
     * 获取数据变更
     * @return {?}
     */
    buildAllEntityChangeDetails() {
        /** @type {?} */
        const changeDetails = [];
        /** @type {?} */
        const entities = this.entityCollection.getAllEntities();
        entities.forEach((/**
         * @param {?} entity
         * @return {?}
         */
        (entity) => {
            if (entity.changes.length === 0) {
                return;
            }
            /** @type {?} */
            const changeDetail = this.changeBuilder.build(entity.changes);
            // 防止空id的变更被提交
            // @todo：此处判断应该放到更底层，临时修复Bug。
            if (changeDetail && changeDetail.ChangeInfo && changeDetail.ChangeInfo.DataId) {
                changeDetails.push(changeDetail);
            }
        }));
        return changeDetails;
    }
    /**
     * 获取单个实体
     * @param {?} id
     * @return {?}
     */
    buildEntityChangeDetailById(id) {
        /** @type {?} */
        const entity = this.entityCollection.getEntityById(id);
        if (entity.changes.length === 0) {
            return null;
        }
        /** @type {?} */
        const changeDetail = this.changeBuilder.build(entity.changes);
        return changeDetail;
    }
    /**
     * 应用数据变更
     * @param {?} changeDetails ChangeDetail数组
     * @return {?}
     */
    handleDataChangeDetails(changeDetails) {
        this.changeHandler.handle(this.entityType, this.entityCollection, changeDetails);
    }
    // #endregion
    /**
     * 清空所有实体
     * @return {?}
     */
    reset() {
        this.entityCollection.clear();
    }
}
if (false) {
    /**
     * ChangeDetail构造器
     * @type {?}
     */
    BefEntityManager.prototype.changeBuilder;
    /**
     * ChangeDetail处理器
     * @type {?}
     */
    BefEntityManager.prototype.changeHandler;
}
export { BefEntityManager };
