// @farris/bef/esm2015/lib/dynamic_bef_repository.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/dynamic_bef_repository.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { DataTypeInfo, EntityCollection } from '@farris/devkit';
import { BefDataService } from './bef_data_service';
import { BefEntityManager } from './bef_entity_manager';
import { BefProxy } from './bef_proxy';
import { BefRepository } from './bef_repository';
import { BE_SERVER_URI_TOKEN } from './tokens';
import { UriService } from './uri_service';
/**
 * @template T
 */
class DynamicBefRepository extends BefRepository {
    // public paginationInfo = {
    //   Entity: {
    //     pageSize: 20,
    //   },
    // };
    /**
     * @param {?} injector
     * @param {?} name
     * @param {?} entityType
     * @param {?} serverUri
     * @param {?} apiUri
     */
    constructor(injector, name, entityType, serverUri, apiUri) {
        super(injector);
        this.name = name;
        this.entityType = entityType;
        this.serverUri = serverUri;
        this.apiUri = apiUri;
        this.apiUri = apiUri;
        // 构造分页属性信息
        /** @type {?} */
        const paginationInfo = {};
        paginationInfo[entityType.typeName] = { pageSize: 20 };
        this.paginationInfo = paginationInfo;
        this.name = name;
        this.entityTypeInfo = new DataTypeInfo(this.entityType);
        this.entityCollection = new EntityCollection(this.entityType);
        // 初始化实体管理器
        this.entityManager = new BefEntityManager(this.entityCollection);
        // 设置proxy属性值
        /** @type {?} */
        const httpClient = injector.get(HttpClient);
        /** @type {?} */
        const uriService = new UriService(serverUri);
        /** @type {?} */
        const proxy = new BefProxy(httpClient, uriService);
        /** @type {?} */
        const baseUri = uriService.extendUri(apiUri);
        proxy.setBaseUri(baseUri);
        this.proxy = proxy;
        this.dataService = new BefDataService(injector, this);
        this.serverUri = injector.get(BE_SERVER_URI_TOKEN);
    }
    /**
     * @param {?} defaultValues
     * @return {?}
     */
    batchAppend(defaultValues) {
        throw new Error('Method not implemented.');
    }
}
if (false) {
    /** @type {?} */
    DynamicBefRepository.prototype.name;
    /** @type {?} */
    DynamicBefRepository.prototype.entityType;
    /** @type {?} */
    DynamicBefRepository.prototype.serverUri;
    /** @type {?} */
    DynamicBefRepository.prototype.apiUri;
}
export { DynamicBefRepository };
