// @farris/bef/esm2015/lib/framework_session_service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/framework_session_service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: Witt
 * @Date: 2018-10-11 20:32:02
 * @Last Modified by: Witt
 * @Last Modified time: 2018-12-13 20:05:46
 */
import { Injectable, Optional } from '@angular/core';
import { SessionService } from '@ecp-caf/caf-common';
class FrameworkSessionService {
    /**
     * @param {?} sessionService
     */
    constructor(sessionService) {
        this.sessionService = sessionService;
    }
    /**
     * 获取用户SessionId
     * @return {?}
     */
    getUserSessionId() {
        return this.sessionService.getUserSessionId();
    }
    /**
     * 获取当前功能菜单的SessionId
     * @param {?=} runtimeContext
     * @return {?}
     */
    getCurrentSessionId(runtimeContext) {
        /** @type {?} */
        let sessionId;
        /** @type {?} */
        const rsi = this.getRuntimeSessionId();
        if (rsi) {
            return rsi;
        }
        if (this.sessionService) {
            if (runtimeContext && runtimeContext.hasOwnProperty('tabId')) {
                /** @type {?} */
                const tabId = runtimeContext.tabId;
                sessionId = this.sessionService.getCurrentSeesionId(tabId);
            }
            else {
                sessionId = this.sessionService.getCurrentSeesionId();
            }
        }
        if (!sessionId) {
            sessionId = this.getUserSessionId();
        }
        return sessionId;
    }
    /**
     * @private
     * @return {?}
     */
    getRuntimeSessionId() {
        /** @type {?} */
        const hash = window.location.hash;
        /** @type {?} */
        const params = this.parse(hash);
        return params && params.rsi || null;
    }
    /**
     * @param {?} queryString
     * @return {?}
     */
    parse(queryString) {
        if (!queryString) {
            return {};
        }
        /** @type {?} */
        const hashes = queryString.slice(queryString.indexOf('?') + 1).split('&');
        return hashes.reduce((/**
         * @param {?} params
         * @param {?} hash
         * @return {?}
         */
        (params, hash) => {
            /** @type {?} */
            const split = hash.indexOf('=');
            /** @type {?} */
            const key = hash.slice(0, split);
            /** @type {?} */
            const val = hash.slice(split + 1);
            return Object.assign(params, { [key]: decodeURIComponent(val) });
        }), {});
    }
}
FrameworkSessionService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FrameworkSessionService.ctorParameters = () => [
    { type: SessionService, decorators: [{ type: Optional }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    FrameworkSessionService.prototype.sessionService;
}
export { FrameworkSessionService };
