// @farris/bef/esm2015/lib/preset_bef_repository.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/preset_bef_repository.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DataTypeInfo, EntityCollection, EntityTypeFactory } from '@farris/devkit';
import { BefEntityManager } from './bef_entity_manager';
import { BefRepository } from './bef_repository';
/**
 * @template T
 */
class PresetBefRepository extends BefRepository {
    /**
     * @param {?} injector
     * @param {?} schema
     */
    constructor(injector, schema) {
        super(injector);
        this.schema = schema;
        /** @type {?} */
        const entityTypeFactory = new EntityTypeFactory();
        this.entityType = (/** @type {?} */ (entityTypeFactory.create(schema)));
        this.entityTypeInfo = new DataTypeInfo(this.entityType);
        this.entityCollection = new EntityCollection(this.entityType);
        // 初始化实体管理器
        this.entityManager = new BefEntityManager(this.entityCollection);
    }
    /**
     * @param {?} defaultValues
     * @return {?}
     */
    batchAppend(defaultValues) {
        throw new Error('Method not implemented.');
    }
}
if (false) {
    /** @type {?} */
    PresetBefRepository.prototype.schema;
}
export { PresetBefRepository };
