// @farris/devkit/esm5/lib/cache/provider/default_cache_provider.js
/**
 * 默认缓存
 * @description
 * 默认缓存使用Map作为缓存容器，存储速度快。但不支持自定义的key对比，对于简单类型的key使用值对比，对于引用类型的key，对比的是引用的地址。
 */
var DefaultCacheProvider = /** @class */ (function () {
    function DefaultCacheProvider() {
        this.buffer = new Map();
    }
    DefaultCacheProvider.prototype.has = function (key) {
        return this.buffer.has(key);
    };
    DefaultCacheProvider.prototype.length = function () {
        return this.buffer.size;
    };
    DefaultCacheProvider.prototype.set = function (value) {
        this.buffer.set(value.key, value);
    };
    DefaultCacheProvider.prototype.get = function (key) {
        return this.buffer.get(key);
    };
    DefaultCacheProvider.prototype.delete = function (key) {
        this.buffer.delete(key);
    };
    DefaultCacheProvider.prototype.clear = function () {
        this.buffer.clear();
    };
    DefaultCacheProvider.prototype.keys = function () {
        return this.buffer.keys();
    };
    DefaultCacheProvider.prototype.values = function () {
        return this.buffer.values();
    };
    return DefaultCacheProvider;
}());
export { DefaultCacheProvider };
