// @farris/devkit/esm5/lib/cache/provider/memory_cache_provider.js
import { CacheKeyCompare } from "../types";
var MemoryCacheProvider = /** @class */ (function () {
    function MemoryCacheProvider(cacheKeyCompare) {
        this.store = new Array();
        this.cacheKeyCompare = cacheKeyCompare || new CacheKeyCompare();
    }
    MemoryCacheProvider.prototype.has = function (key) {
        var _this = this;
        if (this.store.length < 1) {
            return false;
        }
        var index = this.store.findIndex(function (item) { return item && _this.cacheKeyCompare.compare(key, item.key); });
        return index !== -1;
    };
    MemoryCacheProvider.prototype.length = function () {
        return this.store.length;
    };
    MemoryCacheProvider.prototype.set = function (value) {
        this.store.push(value);
    };
    MemoryCacheProvider.prototype.get = function (key) {
        var _this = this;
        if (this.store.length < 1) {
            return undefined;
        }
        return this.store.find(function (item) { return _this.cacheKeyCompare.compare(key, item.key); });
    };
    MemoryCacheProvider.prototype.delete = function (key) {
        var _this = this;
        if (this.store.length < 1) {
            return;
        }
        var index = this.store.findIndex(function (item) { return item && _this.cacheKeyCompare.compare(key, item.key); });
        this.store.splice(index, 1);
    };
    MemoryCacheProvider.prototype.clear = function () {
        this.store.length = 0;
    };
    MemoryCacheProvider.prototype.keys = function () {
        return this.store.keys();
    };
    MemoryCacheProvider.prototype.values = function () {
        return this.store.values();
    };
    return MemoryCacheProvider;
}());
export { MemoryCacheProvider };
