// @gsp-wf/wf-process-model/esm2015/lib/model/extension/condition-expression/Condition.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { MexExpressionImpl } from "../../mex/MexExpressionImpl";
import { ConditionType } from "./ConditionType";
import { ConditionValueVersion } from "./ConditionValueVersion";
export class Condition extends MexExpressionImpl {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = Condition.Clr_Type_ID;
        this._value = '';
        this.valueVersion = ConditionValueVersion.Unsure;
        this.type = ConditionType.CONDITION;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new Condition(this.Model);
    }
    /**
     * @return {?}
     */
    get ConditionType() { return this.type; }
    /**
     * @param {?} value
     * @return {?}
     */
    set ConditionType(value) { this.type = value; }
    /**
     * @return {?}
     */
    get Value() { return this._value; }
    /**
     * @param {?} value
     * @return {?}
     */
    set Value(value) { this._value = value; }
    /**
     * @return {?}
     */
    get ValueVersion() { return this.valueVersion; }
    /**
     * @param {?} value
     * @return {?}
     */
    set ValueVersion(value) { this.valueVersion = value; }
    /**
     * @return {?}
     */
    ToString() {
        return "Condition[value='" + this._value + "'" + ", type=" + this.type + "]";
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.Value = '';
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "ConditionType", this.ConditionType);
        BpmnModelHelper.SetAttributeValue(obj, "Value", this.Value);
        BpmnModelHelper.SetAttributeValue(obj, "ValueVersion", this.ValueVersion);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.ConditionType = obj.ConditionType;
        this.Value = obj.Value;
        this.ValueVersion = obj.ValueVersion;
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const condition = new Condition(this.Model);
        condition.Value = this.Value;
        condition.ValueVersion = this.valueVersion;
        condition.ConditionType = this.ConditionType;
        return condition;
    }
}
Condition.Clr_Type_ID = 'wf.Condition';
if (false) {
    /** @type {?} */
    Condition.Clr_Type_ID;
    /** @type {?} */
    Condition.prototype.ClrTypeID;
    /**
     * @type {?}
     * @private
     */
    Condition.prototype._value;
    /**
     * @type {?}
     * @private
     */
    Condition.prototype.valueVersion;
    /**
     * @type {?}
     * @private
     */
    Condition.prototype.type;
}
BpmnModelHelper.RegisterElementType(new Condition(new BpmnModel()));
