// @gsp-aif/runtime-api/esm2015/lib/aif-push-select-creation-rule/aif-push-select-creation-rule.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { HttpClientModule } from "@angular/common/http";
import { ButtonsModule } from "@progress/kendo-angular-buttons";
import { DropDownsModule } from "@progress/kendo-angular-dropdowns";
import { SplitterModule } from "@progress/kendo-angular-layout";
import { FormsModule, ReactiveFormsModule } from "@angular/forms";
import { HttpService } from "@ecp-caf/caf-common";
import { BsModalService, ModalModule } from '@farris/ui-modal';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { AifPushSelectCreationRuleInternalService } from './service/aif-push-select-creation-rule.internal.service';
import { AifPushSelectCreationRuleComponent } from './aif-push-select-creation-rule/aif-push-select-creation-rule.component';
import { AifPushSelectCreationRuleService } from './service/aif-push-select-creation-rule.service';
import { AifPushSelectCreationRuleModuleConfig } from '../entity/aif-push-select-creation-rule-module-config';
const ɵ0 = {};
export class AifPushSelectCreationRuleModule {
    /**
     * @param {?=} config
     * @return {?}
     */
    static forRoot(config) {
        return {
            "ngModule": AifPushSelectCreationRuleModule,
            "providers": [
                { provide: AifPushSelectCreationRuleModuleConfig, useValue: config || {} },
            ]
        };
    }
}
AifPushSelectCreationRuleModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                    HttpClientModule,
                    ButtonsModule,
                    DropDownsModule,
                    SplitterModule,
                    ModalModule,
                    FarrisDialogModule,
                ],
                providers: [
                    HttpService,
                    BsModalService,
                    { provide: AifPushSelectCreationRuleModuleConfig, useValue: ɵ0 },
                    AifPushSelectCreationRuleInternalService,
                    AifPushSelectCreationRuleService
                ],
                declarations: [
                    AifPushSelectCreationRuleComponent
                ],
                bootstrap: [
                    AifPushSelectCreationRuleComponent
                ],
                entryComponents: [
                    AifPushSelectCreationRuleComponent
                ],
                exports: []
            },] }
];
export { ɵ0 };
