// @farris/component-querycondition/esm2015/lib/querycondition-advanced/querycondition-item/querycondition-item.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/querycondition-advanced/querycondition-item/querycondition-item.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Injector } from '@angular/core';
import { ControlType } from '../../entity/controltype';
import { IdService } from '@farris/ui-common';
import { StateService } from '../service/state.service';
export class QueryconditionItemComponent {
    /**
     * @param {?} injector
     */
    constructor(injector) {
        this.injector = injector;
        this.controlType = ControlType;
        this.fieldCodeOfClearFields = [];
        this.idService = this.injector.get(IdService);
        this.stateSer = this.injector.get(StateService);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    conditionChange() {
        this.stateSer.updateQueryConditionListData();
    }
    /**
     * @param {?} a
     * @return {?}
     */
    getStyle(a) {
        return '';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    buildSmartHelpKey(value) {
        /** @type {?} */
        let prefix = '';
        if (location.hash) {
            /** @type {?} */
            const pathArr = location.hash.split('?');
            prefix = pathArr ? pathArr[0] : '';
        }
        else {
            /** @type {?} */
            const pathArr = location.pathname.split('/');
            prefix = pathArr ? pathArr[pathArr.length - 1] : '';
        }
        return this.idService.encrypt(prefix + '/querysolution/' + value);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    clearFieldsChange(event) {
    }
}
QueryconditionItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'querycondition-item',
                template: "<app-solution-input  \r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.Text && item.visible\"\r\n    [(item)]=\"item\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-input>\r\n\r\n<app-solution-inputgroup\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.InputGroup && item.visible\"\r\n    [(item)]=\"item\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-inputgroup>\r\n\r\n<app-solution-dropdownlist\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.DropDownList && item.visible\"\r\n    [(item)]=\"item\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-dropdownlist>\r\n\r\n<app-solution-smarthelp\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    (clearFieldsChange)=\"clearFieldsChange($event)\"\r\n    *ngIf=\"item.control.getControlType() == controlType.SmartHelp && item.visible\"\r\n    [(item)]=\"item\"\r\n    [instanceId]=\"buildSmartHelpKey(item.fieldCode)\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-smarthelp>\r\n\r\n<app-solution-single-number\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.SingleNumber && item.visible\"\r\n    [(item)]=\"item\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-single-number>\r\n\r\n\r\n\r\n<app-solution-datepicker\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.SingleDate && item.visible\"\r\n    [(item)]=\"item\"\r\n    [controlType]=\"item.control.getControlType()\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-datepicker>\r\n\r\n<app-solution-datepicker\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.DateRange && item.visible\"\r\n    [(item)]=\"item\"\r\n    [controlType]=\"item.control.getControlType()\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-datepicker>\r\n\r\n<app-solution-datepicker\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.SingleYear && item.visible\"\r\n    [(item)]=\"item\"\r\n    [controlType]=\"item.control.getControlType()\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-datepicker>\r\n\r\n<app-solution-datepicker\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.SingleMonth && item.visible\"\r\n    [(item)]=\"item\"\r\n    [controlType]=\"item.control.getControlType()\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-datepicker>\r\n\r\n<app-solution-datepicker\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.MonthRange && item.visible\"\r\n    [(item)]=\"item\"\r\n    [controlType]=\"item.control.getControlType()\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-datepicker>\r\n\r\n<app-solution-datepicker\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.SingleDateTime && item.visible\"\r\n    [(item)]=\"item\"\r\n    [controlType]=\"item.control.getControlType()\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-datepicker>\r\n\r\n<app-solution-datepicker\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.DateTimeRange && item.visible\"\r\n    [(item)]=\"item\"\r\n    [controlType]=\"item.control.getControlType()\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-datepicker>\r\n\r\n<app-solution-combolookup\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.ComboLookUp && item.visible\"\r\n    [(item)]=\"item\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-combolookup>\r\n\r\n<app-solution-checkbox\r\n    [ngClass]=\"getStyle(item)\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.BoolCheck && item.visible\"\r\n    [(item)]=\"item\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-checkbox>\r\n\r\n<app-solution-radio\r\n    [ngClass]=\"getStyle(item)\"\r\n    class=\"\"\r\n    [changeFields]=\"fieldCodeOfClearFields\"\r\n    (conditionChange)=\"conditionChange()\"\r\n    *ngIf=\"item.control.getControlType() == controlType.Radio && item.visible\"\r\n    [(item)]=\"item\"\r\n    [hideLabel]=\"true\">\r\n</app-solution-radio>",
                styles: [""]
            }] }
];
/** @nocollapse */
QueryconditionItemComponent.ctorParameters = () => [
    { type: Injector }
];
QueryconditionItemComponent.propDecorators = {
    item: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    QueryconditionItemComponent.prototype.controlType;
    /** @type {?} */
    QueryconditionItemComponent.prototype.item;
    /** @type {?} */
    QueryconditionItemComponent.prototype.fieldCodeOfClearFields;
    /**
     * @type {?}
     * @private
     */
    QueryconditionItemComponent.prototype.idService;
    /**
     * @type {?}
     * @private
     */
    QueryconditionItemComponent.prototype.stateSer;
    /**
     * @type {?}
     * @private
     */
    QueryconditionItemComponent.prototype.injector;
}
