// @farris/command-services/esm5/lib/command-service.js
import { Injectable } from '@angular/core';
import { filter, take } from 'rxjs/operators';
import { AppContext, ViewModel } from '@farris/devkit';
/**
 * 命令服务
 * @scope FrameConmponent
 */
var CommandService = /** @class */ (function () {
    function CommandService(viewModel, appContext) {
        this.viewModel = viewModel;
        this.appContext = appContext;
    }
    Object.defineProperty(CommandService.prototype, "context", {
        /**
         * context取值方法
         */
        get: function () {
            return this.innerContext;
        },
        /**
         * context设值方法
         */
        set: function (context) {
            this.innerContext = context;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 执行命令
     * @param commandName 命令名称
     * @param frameId 框架ID
     */
    CommandService.prototype.execute = function (commandName, frameId) {
        if (!commandName || commandName === '' || commandName === 'undefined') {
            return;
        }
        var viewModel = this.viewModel;
        if (frameId) {
            var targetContext = this.appContext.frameContextManager.getFrameContextById(frameId);
            viewModel = targetContext.viewModel;
        }
        return viewModel[commandName]();
    };
    /**
     * 等待BeSession创建
     * @returns 最新创建出来的BeSession
     */
    CommandService.prototype.waitForBeSession = function () {
        var frameContext = this.context.frameContext;
        var befRepository = frameContext.repository;
        var sessionService = befRepository.restService.sessionService;
        return sessionService.getBeSessionId().pipe(filter(function (beSessionId) {
            return beSessionId !== null;
        }), take(1));
    };
    /**
     * 提取某个任务的执行结果
     * @param taskName 任务名称
     * @reutrn 返回taskName对应任务的执行结果
     */
    CommandService.prototype.extractResult = function (taskName) {
        var results = this.context.results;
        return results[taskName];
    };
    /**
     * 暂停当前组件的行切换事件抛出
     */
    CommandService.prototype.suspendFrameContextRowSelectedEvent = function (frameId) {
        var frameContext = this.viewModel.frameContext;
        if (frameId) {
            frameContext = this.viewModel.frameContext.appContext.frameContextManager.getFrameContextById(frameId);
        }
        if (frameContext) {
            frameContext.bindingData.rowSelectedEventSuspend = true;
        }
    };
    CommandService.prototype.resumeFrameContextRowSelectedEvent = function (frameId) {
        var frameContext = this.viewModel.frameContext;
        if (frameId) {
            frameContext = this.viewModel.frameContext.appContext.frameContextManager.getFrameContextById(frameId);
        }
        if (frameContext) {
            frameContext.bindingData.rowSelectedEventSuspend = false;
        }
    };
    CommandService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CommandService.ctorParameters = function () { return [
        { type: ViewModel },
        { type: AppContext }
    ]; };
    return CommandService;
}());
export { CommandService };
