// @farris/command-services/esm5/lib/dirty-checking.service.js
import { ApplicationRef, ChangeDetectorRef, Injectable, Injector } from "@angular/core";
/**
 * angular 脏值检查服务
 */
var DirtyCheckingService = /** @class */ (function () {
    function DirtyCheckingService(injector, cd, appRef) {
        this.injector = injector;
        this.cd = cd;
        this.appRef = appRef;
    }
    /**
     * Checks this view and its children.
     * @param global 是否全局检查
     */
    DirtyCheckingService.prototype.detectChanges = function (global) {
        global = (global === true || global === 'true') ? true : false;
        if (global && this.appRef) {
            this.appRef.tick();
        }
        else if (this.cd) {
            this.cd.detectChanges();
        }
    };
    /**
     * 在指定时间（ms）后触发Angular脏值检查
     * @param duration 超时时间(ms)
     * @param global 是否全局检查
     */
    DirtyCheckingService.prototype.detectChangesAfter = function (duration, global) {
        var _this = this;
        global = (global === true || global === 'true') ? true : false;
        if (isNaN(duration)) {
            return;
        }
        if (global && this.appRef) {
            window.setTimeout(function () {
                _this.appRef.tick();
            }, duration);
        }
        else if (this.cd) {
            window.setTimeout(function () {
                _this.cd.detectChanges();
            }, duration);
        }
    };
    DirtyCheckingService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DirtyCheckingService.ctorParameters = function () { return [
        { type: Injector },
        { type: ChangeDetectorRef },
        { type: ApplicationRef }
    ]; };
    return DirtyCheckingService;
}());
export { DirtyCheckingService };
