// @farris/command-services/esm5/lib/exception.handler.js
/*
 * @Author: aalizzwell
 * @Date: 2019-05-30 10:35:09
 * @Last Modified by: aalizzwell
 * @Last Modified time: 2019-06-15 19:22:13
 */
import { Injectable, Optional } from '@angular/core';
import { LanguageService } from './languag.service';
import { FormErrorService } from './error/form-error.service';
import { AppContext } from '@farris/devkit';
/**
 * 默认异常处理器
 * @scope 没有提供对应Provider
 */
var ExceptionHandler = /** @class */ (function () {
    function ExceptionHandler(formErrorService, languageService, applicationContext) {
        this.formErrorService = formErrorService;
        this.languageService = languageService;
        this.applicationContext = applicationContext;
        this.appContext = null;
        if (this.languageService == null) {
            this.languageService = LanguageService.getInstance();
        }
    }
    ExceptionHandler.prototype.setContext = function (applicationContext) {
        this.appContext = applicationContext;
    };
    ExceptionHandler.prototype.handle = function (error) {
        if (!!this.formErrorService) {
            this.formErrorService.exception(this.languageService.loadFailed, error);
        }
    };
    ExceptionHandler.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ExceptionHandler.ctorParameters = function () { return [
        { type: FormErrorService },
        { type: LanguageService, decorators: [{ type: Optional }] },
        { type: AppContext }
    ]; };
    return ExceptionHandler;
}());
export { ExceptionHandler };
