// @farris/command-services/esm5/lib/form-wizard.service.js
import { Injectable } from '@angular/core';
import { WizardService, WizardEventType } from '@farris/ui-wizard';
var FormWizardService = /** @class */ (function () {
    function FormWizardService(wizardSer) {
        this.wizardSer = wizardSer;
    }
    FormWizardService.prototype.preStep = function () {
        var currentPageId = this.wizardSer.getCurrentPageId();
        this.wizardSer.updateWizardState(WizardEventType.Prev, currentPageId);
    };
    FormWizardService.prototype.nextStep = function () {
        var currentPageId = this.wizardSer.getCurrentPageId();
        this.wizardSer.updateWizardState(WizardEventType.Next, currentPageId);
    };
    FormWizardService.prototype.cancelWizard = function () {
        this.wizardSer.cancelWizard();
    };
    FormWizardService.prototype.finishWizard = function () {
        this.wizardSer.finishWizard();
    };
    FormWizardService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FormWizardService.ctorParameters = function () { return [
        { type: WizardService }
    ]; };
    return FormWizardService;
}());
export { FormWizardService };
