// @farris/command-services/esm5/lib/form-attention.service.js
import { Injectable, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { EMPTY } from "rxjs";
import { AttentionService } from '@farris/ui-tooltip';
var FormAttentionService = /** @class */ (function () {
    function FormAttentionService(injector, frameContext, attentionService) {
        this.injector = injector;
        this.frameContext = frameContext;
        this.attentionService = attentionService;
    }
    /**
     * 闪烁按钮
     * @param ids id字符串或数组
     * @param duration 时间间隔
     * @returns
     */
    FormAttentionService.prototype.blink = function (ids, duration) {
        if (typeof ids === 'string') {
            ids = ids.split(',').filter(function (p) { return p; });
        }
        if (!Array.isArray(ids) || ids.length < 1) {
            return EMPTY;
        }
        if (typeof duration === 'string') {
            duration = duration.trim();
        }
        if (!duration) {
            duration = 600;
        }
        duration = parseInt(duration);
        if (isNaN(duration) || duration <= 0) {
            duration = 600;
        }
        var elements = ids.map(function (item) {
            return {
                idOrEl: item,
            };
        });
        this.attentionService.catchAttention(elements, duration);
    };
    FormAttentionService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FormAttentionService.ctorParameters = function () { return [
        { type: Injector },
        { type: FrameContext },
        { type: AttentionService }
    ]; };
    return FormAttentionService;
}());
export { FormAttentionService };
