// @farris/command-services/esm5/lib/menu-state.service.js
import { Injectable } from '@angular/core';
// tslint:disable: no-unused-expression
/**
 * 菜单状态服务
 * @scope FormModule
 */
var MenuStateService = /** @class */ (function () {
    function MenuStateService() {
        this.COMMAND_SERVICE_LINK_STORAGE_KEY = 'COMMAND_SERVICE_LINK_STORAGE_KEY';
    }
    MenuStateService.prototype.getState = function () {
        var data = !!sessionStorage && sessionStorage.getItem(this.COMMAND_SERVICE_LINK_STORAGE_KEY) || null;
        if (!!data) {
            return JSON.parse(data);
        }
        return {};
    };
    MenuStateService.prototype.setState = function (links) {
        !!sessionStorage && sessionStorage.setItem(this.COMMAND_SERVICE_LINK_STORAGE_KEY, JSON.stringify(links));
    };
    MenuStateService.prototype.addMenuState = function (parentMenuId, childMenuId, status) {
        if (status === void 0) { status = true; }
        if (!childMenuId) {
            return;
        }
        var state = this.getState();
        var isParentExist = !!state && state.hasOwnProperty(parentMenuId);
        var childItem = { id: childMenuId, status: status };
        if (isParentExist) {
            var parent_1 = state[parentMenuId];
            var isChildItemExist = parent_1.find(function (item) { return item.id === childMenuId; });
            if (!isChildItemExist) {
                state[parentMenuId].push(childItem);
            }
        }
        else {
            state[parentMenuId] = [childItem];
        }
        this.setState(state);
    };
    MenuStateService.prototype.getMenuState = function (parentMenuId, childMenuId) {
        var state = this.getState();
        if (state === null) {
            return null;
        }
        var isParentExist = state.hasOwnProperty(parentMenuId);
        if (!isParentExist) {
            return null;
        }
        var parentMenu = state[parentMenuId];
        var childMenu = parentMenu.filter(function (item) { return item.status === true; });
        if (typeof childMenuId === 'undefined' || childMenuId.length < 1) {
            return childMenu;
        }
        return parentMenu.filter(function (item) { return item.status && item.id === childMenuId; });
    };
    MenuStateService.prototype.updateMenuState = function (childMenuId, status) {
        var state = this.getState();
        if (state === null) {
            return;
        }
        Object.keys(state).forEach(function (key) {
            var childMenu = state[key].find(function (item) { return item.id === childMenuId; });
            if (!!childMenu) {
                childMenu.status = status;
            }
        });
        this.setState(state);
    };
    MenuStateService.prototype.removeMenu = function (menuId) {
        this.removeParentMenu(menuId);
        this.removeChildMenu(menuId);
    };
    MenuStateService.prototype.removeParentMenu = function (menuId) {
        var state = this.getState();
        if (state === null) {
            return null;
        }
        Object.keys(state).forEach(function (item) {
            if (item === menuId && state[menuId].length < 1) {
                delete state[item];
            }
        });
        this.setState(state);
    };
    MenuStateService.prototype.removeChildMenu = function (menuId) {
        var state = this.getState();
        if (state === null) {
            return;
        }
        Object.keys(state).forEach(function (parentMenuId) {
            var childMenus = state[parentMenuId];
            if (childMenus && childMenus.length > 0) {
                childMenus.forEach(function (child, index) {
                    if (child.id === menuId) {
                        childMenus.splice(index, 1);
                    }
                });
            }
        });
        this.setState(state);
    };
    MenuStateService.decorators = [
        { type: Injectable }
    ];
    return MenuStateService;
}());
export { MenuStateService };
