// @farris/command-services/esm5/lib/param.service.js
import { Injectable } from '@angular/core';
import { QuerystringService } from './querystring';
import { RuntimeFrameworkService } from './rtf-service';
import { of, Subject } from 'rxjs';
import { switchMap } from 'rxjs/operators';
// tslint:disable: no-string-literal
/**
 * 参数服务
 * @scope FormModule
 */
var ParamService = /** @class */ (function () {
    function ParamService(querystringService, runtimeFrameworkService) {
        this.querystringService = querystringService;
        this.runtimeFrameworkService = runtimeFrameworkService;
    }
    Object.defineProperty(ParamService.prototype, "params", {
        /**
         * 适配获取所有参数
         */
        get: function () {
            var _this = this;
            // 先从hash中获取参数
            var hash = window.location.hash;
            var params = this.querystringService.parse(hash);
            /*
            const appMode = params.appMode;
            // appMode没有定义，认为是老表单,appMode = mdi 或 spa
            if (typeof appMode === 'undefined' || appMode === null) {
                return of(params);
            }
            // appMode存在，但为空，也认为是老表单
            if (appMode.length < 1) {
                return of(params);
            }
            // 新表单，使用iframe模式
            if (appMode.toLowerCase() === 'mdi') {
                return of(params);
            }
            */
            // 获取tabId，最新版本一定有tabId，如果取不到tabId则认为是老表单，直接从url获取参数
            var tabId = params.tabId;
            if (!tabId) {
                // throw new Error('TabId can`t be empty!');
                return of(params);
            }
            var subject = new Subject();
            // 管道参数e可能为object、Map类型
            this.runtimeFrameworkService.addEventListener(tabId, function (e) {
                var map = {};
                if (e instanceof Map || (e && typeof (e['get']) === 'function' && typeof e['entries'] === 'function')) {
                    map = _this.parseMapParams(e);
                }
                var args = {};
                if (e instanceof Map || (e && typeof (e['get']) === 'function' && typeof e['entries'] === 'function')) {
                    args = new Map(e);
                }
                else {
                    args = Object.assign({}, e);
                }
                setTimeout(function () {
                    subject.next(Object.assign({}, args, map, params));
                }, 0);
            }, params);
            return subject.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * 解析map类型的参数
     * @param params params
     */
    ParamService.prototype.parseMapParams = function (params) {
        var map = {};
        map['WEB_FORM_ROUTE_PARAMS'] = decodeURIComponent(params.get('WEB_FORM_ROUTE_PARAMS'));
        params.forEach(function (value, key) {
            if (key !== 'WEB_FORM_ROUTE_PARAMS') {
                map[key] = value;
            }
        });
        return map;
    };
    /**
     * 获取param参数
     * @param param key
     */
    ParamService.prototype.get = function (param) {
        return this.params.pipe(switchMap(function (options) {
            if (!!options && options.hasOwnProperty(param)) {
                return of(options.param);
            }
            else {
                return of(undefined);
            }
        }));
    };
    /**
     * 参数转为对象
     */
    ParamService.prototype.parse = function () {
        return this.params.pipe(switchMap(function (options) {
            return of(options);
        }));
    };
    ParamService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ParamService.ctorParameters = function () { return [
        { type: QuerystringService },
        { type: RuntimeFrameworkService }
    ]; };
    return ParamService;
}());
export { ParamService };
