// @farris/command-services/esm5/lib/print.service.js
import { Injectable, Optional } from '@angular/core';
import { EMPTY } from 'rxjs';
import { FormMessageService } from './form-message.service';
import { LanguageService } from './languag.service';
import { CloudprintService, OutputType, FileType, PrintType } from '@gsp-svc/cloudprint';
import { FormNotifyService } from './form-notify.service';
// tslint:disable: unified-signatures
// tslint:disable: max-line-length
/**
 * 打印服务
 * @Scope FrameComponent
 */
var PrintService = /** @class */ (function () {
    /**
     * 构造函数
     */
    function PrintService(msgService, languageService, printService, formNotifyService) {
        this.msgService = msgService;
        this.languageService = languageService;
        this.printService = printService;
        this.formNotifyService = formNotifyService;
    }
    /**
     * 打印单据
     * @param beMetaId BE元数据标识
     * @param bizBillId 业务单据标识
     */
    PrintService.prototype.printSingle = function (beMetaId, bizBillId) {
        if (!bizBillId) {
            // this.msgService.error(this.languageService.unallowEmptyBizBillId);
            this.showWarning(this.languageService.unallowEmptyBizBillId);
            return EMPTY;
        }
        return this.printArray(beMetaId, [bizBillId]);
    };
    /**
     * 打印单据（通过id或ids）
     * @param beMetaId be元数据Id
     * @param ids 单据id或ids
     */
    PrintService.prototype.printByIds = function (beMetaId, ids) {
        if (!ids) {
            // this.msgService.error(this.languageService.unallowEmptyBizBillId);
            this.showWarning(this.languageService.unallowEmptyBizBillId);
            return EMPTY;
        }
        var dataIds = ids.split(',').filter(function (item) { return item; });
        return this.printArray(beMetaId, dataIds);
    };
    /**
     * 打印单据（带维度）
     * @param beMetaId be元数据Id
     * @param ids 单据id或ids
     * @param dim1 维度1值
     * @param dim2 维度2值
     * @param billCategoryId 单据类型Id
     */
    PrintService.prototype.printByIdsWithDimension = function (beMetaId, ids, dim1, dim2, billCategoryId) {
        if (!ids) {
            // this.msgService.error(this.languageService.unallowEmptyBizBillId);
            this.showWarning(this.languageService.unallowEmptyBizBillId);
            return EMPTY;
        }
        var dataIds = ids.split(',').filter(function (item) { return item; });
        return this.printArray(beMetaId, dataIds, dim1, dim2, billCategoryId);
    };
    /**
     * 打印多个单据
     * @param beMetaId BE元数据标识
     * @param dataIds 业务单据标识数组
     * @param dim1 维度1
     * @param dim2 维度2
     * @param billCategoryId 业务单据类型Id
     */
    PrintService.prototype.printArray = function (beMetaId, dataIds, dim1, dim2, billCategoryId) {
        if (!dataIds || dataIds.length === 0) {
            // this.msgService.error(this.languageService.unallowEmptyBizBillId);
            this.showWarning(this.languageService.unallowEmptyBizBillId);
            return EMPTY;
        }
        var sourceOptions = this.buildSourceOptions({
            dataIds: dataIds,
            sourceId: beMetaId
        });
        var outputOptions = this.buildOutputOptions();
        if (typeof dim1 !== 'undefined') {
            sourceOptions.FirstDimensionVal = dim1;
        }
        if (typeof dim2 !== 'undefined') {
            sourceOptions.SecondDimensionVal = dim2;
        }
        if (typeof billCategoryId !== 'undefined') {
            sourceOptions.billCategoryId = billCategoryId;
        }
        return this.printService.outputBEData(sourceOptions, outputOptions, 'tab');
    };
    /**
     * 按照BE取数方式批量打印单据
     * @param beMetaId BE元数据标识
     * @param filterCondition 过滤条件
     * @param sortCondition 排序条件
     * @param includeChildData 包含子表数据
     */
    PrintService.prototype.printMulti = function (beMetaId, filterCondition, sortCondition, includeChildData) {
        if (includeChildData === void 0) { includeChildData = true; }
        var entryFilter = { 'isUsePagination': false, 'filterConditions': [], 'sortConditions': [], 'pagination': null };
        if (filterCondition) {
            // 统一纠正最后一个过滤条件的Relation
            var filters = JSON.parse(filterCondition);
            if (filters && filters.length > 0) {
                filters[filters.length - 1].Relation = 0;
            }
            entryFilter.filterConditions = filters;
        }
        if (sortCondition) {
            entryFilter.sortConditions = JSON.parse(sortCondition);
        }
        // sfo:SourceFilterOptions
        var sourceFilterOptions = this.buildSourceFilterOptions({ sourceId: beMetaId, entryFilter: entryFilter, includeChildData: includeChildData });
        var outputOptions = this.buildOutputOptions();
        return this.printService.outputBEDataWithFilter(sourceFilterOptions, outputOptions, 'tab');
    };
    /**
     * 按照BE取数方式批量打印单据(带维度)
     * @param beMetaId BE元数据标识
     * @param filterCondition 过滤条件
     * @param sortCondition 排序条件
     * @param dim1 维度1value
     * @param dim2 维度2value
     * @param billCategoryId 业务单据类型Id
     * @param includeChildData 包含子表数据
     */
    PrintService.prototype.printMultiWithDimension = function (beMetaId, filterCondition, sortCondition, dim1, dim2, billCategoryId, includeChildData) {
        if (includeChildData === void 0) { includeChildData = true; }
        var entryFilter = { 'isUsePagination': false, 'filterConditions': [], 'sortConditions': [], 'pagination': null };
        if (filterCondition) {
            // 统一纠正最后一个过滤条件的Relation
            var filters = JSON.parse(filterCondition);
            if (filters && filters.length > 0) {
                filters[filters.length - 1].Relation = 0;
            }
            entryFilter.filterConditions = filters;
        }
        if (sortCondition) {
            entryFilter.sortConditions = JSON.parse(sortCondition);
        }
        var sfo = this.buildSourceFilterOptions({ sourceId: beMetaId, entryFilter: entryFilter, includeChildData: includeChildData });
        if (typeof dim1 !== 'undefined') {
            sfo.FirstDimensionVal = dim1;
        }
        if (typeof dim2 !== 'undefined') {
            sfo.SecondDimensionVal = dim2;
        }
        if (typeof billCategoryId !== 'undefined') {
            sfo.billCategoryId = billCategoryId;
        }
        var outputOptions = this.buildOutputOptions();
        return this.printService.outputBEDataWithFilter(sfo, outputOptions, 'tab');
    };
    /**
     * 构造SourceOptions
     * @param options options
     */
    PrintService.prototype.buildSourceOptions = function (options) {
        var so = {
            DataIds: options && options.dataIds || undefined,
            SourceId: options && options.sourceId || undefined,
            FirstDimensionVal: options && options.dim1 || undefined,
            SecondDimensionVal: options && options.dim2 || undefined,
            RetrieveParam: options && options.retrieveParam || undefined,
            FormatId: options && options.formatId || undefined,
            billCategoryId: options && options.billCategoryId || undefined,
            ServiceUnit: options && options.serviceUnit || undefined,
            currentPage: options && options.currentPage || undefined,
            pageRowCount: options && options.pageRowCount || undefined,
            queryType: options && options.queryType || undefined,
            queryServiceId: options && options.queryServiceId || undefined,
            queryParam: options && options.queryParam || undefined
        };
        return so;
    };
    /**
     * 构造OutputOptions
     * @param options options
     */
    PrintService.prototype.buildOutputOptions = function (options) {
        var oo = {
            OutputType: options && options.outputType || OutputType.PRINT,
            FileType: options && options.fileType || FileType.Html5,
            Path: options && options.path || undefined,
            DeviceId: options && options.deviceId || undefined,
            printJob: options && options.printJob || undefined,
            printerName: options && options.printerName || undefined,
            printSetting: options && options.printSetting || undefined,
            printType: options && options.printType || PrintType.Form
        };
        return oo;
    };
    /**
     * 构造SourceFilterOptions
     * @param options options
     */
    PrintService.prototype.buildSourceFilterOptions = function (options) {
        var entryFilter = { 'isUsePagination': false, 'filterConditions': [], 'sortConditions': [], 'pagination': null };
        var sfo = {
            SourceId: options.sourceId,
            EntityFilter: options && options.entryFilter || entryFilter,
            FirstDimensionVal: options && options.dim1 || undefined,
            SecondDimensionVal: options && options.dim2 || undefined,
            FormatId: options && options.formatId || undefined,
            ServiceUnit: options && options.serviceUnit || undefined,
            billCategoryId: options && options.billCategoryId || undefined,
            currentPage: options && options.currentPage || undefined,
            pageRowCount: options && options.pageRowCount || undefined,
            queryParam: options && options.queryParam || undefined,
            queryServiceId: options && options.queryServiceId || undefined,
            queryType: options && options.queryType || undefined,
            includeChildData: options && options.hasOwnProperty('includeChildData') ? options.includeChildData : true
        };
        return sfo;
    };
    /**
     * 展示错误消息
     * @param message 错误消息
     */
    PrintService.prototype.showWarning = function (message) {
        if (this.notifyService) {
            this.notifyService.warning(message, { hideTitle: true });
        }
        else if (this.msgService) {
            this.msgService.error(message);
        }
    };
    Object.defineProperty(PrintService.prototype, "notifyService", {
        get: function () {
            if (this.formNotifyService) {
                return this.formNotifyService;
            }
            else if (this.injector) {
                return this.injector.get(FormNotifyService, null);
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PrintService.prototype, "commandContext", {
        get: function () {
            return this['context'] || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PrintService.prototype, "frameContext", {
        get: function () {
            return this.commandContext && this.commandContext.frameContext || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PrintService.prototype, "injector", {
        get: function () {
            return this.frameContext && this.frameContext.injector || null;
        },
        enumerable: true,
        configurable: true
    });
    PrintService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    PrintService.ctorParameters = function () { return [
        { type: FormMessageService },
        { type: LanguageService },
        { type: CloudprintService },
        { type: FormNotifyService, decorators: [{ type: Optional }] }
    ]; };
    return PrintService;
}());
export { PrintService };
