// @farris/command-services/esm5/lib/sidebar-service.js
import { Injectable, Optional } from '@angular/core';
import { of } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { FarrisSidebarService } from '@farris/ui-sidebar';
import { FrameContext } from '@farris/devkit';
import { FormMessageService } from './form-message.service';
import { LanguageService } from './languag.service';
/**
 * 侧边栏服务
 * @scope FrameComponent
 */
var SidebarService = /** @class */ (function () {
    function SidebarService(frameContext, sidebarUIService, messageService, languageService) {
        this.frameContext = frameContext;
        this.sidebarUIService = sidebarUIService;
        this.messageService = messageService;
        this.languageService = languageService;
        if (!this.languageService) {
            this.languageService = LanguageService.getInstance();
        }
    }
    Object.defineProperty(SidebarService.prototype, "repository", {
        get: function () {
            return this.frameContext.repository;
        },
        enumerable: true,
        configurable: true
    });
    SidebarService.prototype.openSidebar = function () {
        this.sidebarUIService.sendIsOpen(true);
    };
    SidebarService.prototype.closeSidebar = function () {
        this.sidebarUIService.sendIsOpen(false);
    };
    /**
     * 关闭前先取消变更
     */
    SidebarService.prototype.confirmBeforeClosingSidebar = function () {
        // 检查是否有变更，没有变更直接关闭
        var ifChangesExist = this.repository.entityManager.checkAllEntityChanges();
        if (!ifChangesExist) {
            return of(true);
        }
        // 确认是否关闭
        var confirm$ = this.messageService.question(this.languageService['exitWithoutSave']);
        var result$ = confirm$.pipe(switchMap(function (confirmResult) {
            if (confirmResult === false) {
                return of(false);
            }
            else {
                return of(true);
            }
        }));
        return result$;
    };
    /**
     * 继续关闭侧边栏
     */
    SidebarService.prototype.continueClosingSidebar = function () {
        return of(true);
    };
    /**
     * 阻止侧边栏关闭
     */
    SidebarService.prototype.stopClosingSidebar = function () {
        return of(false);
    };
    SidebarService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SidebarService.ctorParameters = function () { return [
        { type: FrameContext },
        { type: FarrisSidebarService },
        { type: FormMessageService },
        { type: LanguageService, decorators: [{ type: Optional }] }
    ]; };
    return SidebarService;
}());
export { SidebarService };
