// @farris/devkit/esm2015/lib/listener/ui_state_change_listener.js
import { Injectable, Injector, Inject } from '@angular/core';
// import { FRAME_ID, NAMESPACE } from '../frame/index';
import { UIState } from '../ui-state/index';
import { ChangeListener } from './change_listener';
import { Expression } from '../expression/index';
import { BindingData } from '../binding-data/index';
import { FRAME_ID, NAMESPACE } from '../frame/tokens';
/**
 * 监听UIState变更
 */
class UIStateChangeListener extends ChangeListener {
    constructor(injector, uiState, namespace, frameId, bindingData) {
        super();
        this.injector = injector;
        this.uiState = uiState;
        this.namespace = namespace;
        this.frameId = frameId;
        this.bindingData = bindingData;
        this.registerEvent();
    }
    buildEventPath(change) {
        return null;
    }
    registerEvent() {
        if (this.uiState && this.uiState.changes) {
            this.uiState.changes.subscribe((change) => {
                const modification = {
                    ns: this.namespace,
                    path: [change.field],
                    type: Expression.EventType.ValueChanged,
                    value: change.value,
                    source: Expression.EventSource.State,
                    frameId: this.frameId
                };
                // console.log("UIStateChangeListener", modification);
                this.subject.next(modification);
            });
        }
    }
}
UIStateChangeListener.decorators = [
    { type: Injectable }
];
/** @nocollapse */
UIStateChangeListener.ctorParameters = () => [
    { type: Injector },
    { type: UIState },
    { type: undefined, decorators: [{ type: Inject, args: [NAMESPACE,] }] },
    { type: String, decorators: [{ type: Inject, args: [FRAME_ID,] }] },
    { type: BindingData }
];
export { UIStateChangeListener };
