// @gsp-bef/gsp-cm-metadata/esm2015/lib/common-model/src/Element/ElementCodeRuleConfig.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { GSPBillCodeGenerateType } from './GSPBillCodeGenerateType';
import { GSPBillCodeGenerateOccasion } from './GSPBillCodeGenerateOccasion';
/**
 * 编码规则配置
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 编码规则配置
 */
export class ElementCodeRuleConfig {
    constructor() {
        this.CanBillCode = false;
        this.BillCodeID = '';
        this.BillCodeName = '';
        this.CodeGenerateType = GSPBillCodeGenerateType.none;
        this.CodeGenerateOccasion = GSPBillCodeGenerateOccasion.SystemProcess;
    }
}
if (false) {
    /**
     * 是否单据编号
     * @type {?}
     */
    ElementCodeRuleConfig.prototype.CanBillCode;
    /**
     * 单据编号Id
     * @type {?}
     */
    ElementCodeRuleConfig.prototype.BillCodeID;
    /**
     * 单据编号名称
     * @type {?}
     */
    ElementCodeRuleConfig.prototype.BillCodeName;
    /**
     * 代码生成类型
     * @type {?}
     */
    ElementCodeRuleConfig.prototype.CodeGenerateType;
    /**
     * 编码生成时机
     * @type {?}
     */
    ElementCodeRuleConfig.prototype.CodeGenerateOccasion;
}
