// @farris/approval-comments/esm2015/lib/approval-comments.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/approval-comments.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ApiService } from '../utils/api.service';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { FormsModule } from '@angular/forms';
import { FEPageModalModule } from '@farris/extend-page-modal';
import { ComboListModule } from '@farris/ui-combo-list';
import { FarrisFormsModule } from '@farris/ui-forms';
import { InputGroupModule } from '@farris/ui-input-group';
import { NotifyModule } from '@farris/ui-notify';
import { RateModule } from '@farris/ui-rate';
import { SwitchModule } from '@farris/ui-switch';
import { ApprovalCommentEditorComponent } from './approval-comment-editor/approval-comment-editor.component';
import { ApprovalCommentComponent } from './approval-comments.component';
export class ApprovalCommentsModule {
}
ApprovalCommentsModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    ApprovalCommentComponent,
                    ApprovalCommentEditorComponent
                ],
                imports: [
                    CommonModule,
                    FormsModule,
                    FarrisFormsModule,
                    DragDropModule,
                    FEPageModalModule,
                    NotifyModule,
                    RateModule,
                    SwitchModule,
                    ComboListModule,
                    InputGroupModule
                ],
                exports: [ApprovalCommentComponent, ApprovalCommentEditorComponent],
                entryComponents: [ApprovalCommentComponent, ApprovalCommentEditorComponent],
                providers: [ApiService]
            },] }
];
