// @gsp-wf/wf-process-editor/esm5/lib/wf-process-editor-routing.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { TranslateResolveService } from './service/translate-resolve.service';
import { WfProcessEditorComponent } from './wf-process-editor.component';
/** @type {?} */
var routes = [
    {
        path: 'WfProcessEditor', component: WfProcessEditorComponent,
        children: [],
        resolve: {
            'translate': TranslateResolveService
        }
    }
];
var WfProcessEditorRoutingModule = /** @class */ (function () {
    function WfProcessEditorRoutingModule() {
    }
    WfProcessEditorRoutingModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        RouterModule.forChild(routes)
                    ],
                    exports: [
                        RouterModule
                    ]
                },] }
    ];
    return WfProcessEditorRoutingModule;
}());
export { WfProcessEditorRoutingModule };
