// @ecp-caf/common-structure/esm5/lib/define/data-type/primitive-type/primitive-type-kind.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var PrimitiveTypeKind = {
    // String = 0,
    // Bool = 1,
    // Int16 = 2,
    // Int32 = 3,
    // Int64 = 4,
    // Float = 5,
    // Double = 6,
    // Decimal = 7,
    // DateTime = 8,
    // TimeStamp = 9,
    // Byte=10 ,
    String: 0,
    Bool: 1,
    Int16: 2,
    Int32: 3,
    Int64: 4,
    Float: 5,
    Double: 6,
    Decimal: 7,
    DateTime: 8,
    TimeStamp: 9,
    Byte: 10,
};
export { PrimitiveTypeKind };
PrimitiveTypeKind[PrimitiveTypeKind.String] = 'String';
PrimitiveTypeKind[PrimitiveTypeKind.Bool] = 'Bool';
PrimitiveTypeKind[PrimitiveTypeKind.Int16] = 'Int16';
PrimitiveTypeKind[PrimitiveTypeKind.Int32] = 'Int32';
PrimitiveTypeKind[PrimitiveTypeKind.Int64] = 'Int64';
PrimitiveTypeKind[PrimitiveTypeKind.Float] = 'Float';
PrimitiveTypeKind[PrimitiveTypeKind.Double] = 'Double';
PrimitiveTypeKind[PrimitiveTypeKind.Decimal] = 'Decimal';
PrimitiveTypeKind[PrimitiveTypeKind.DateTime] = 'DateTime';
PrimitiveTypeKind[PrimitiveTypeKind.TimeStamp] = 'TimeStamp';
PrimitiveTypeKind[PrimitiveTypeKind.Byte] = 'Byte';
