// @farris/ui-notify/esm2015/lib/notify.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FarrisCommonModule } from '@farris/ui-common';
import { NotifyComponent } from './notify.component';
import { NotifyContainerComponent } from './notify-container.component';
import { NotifyService } from './notify.service';
import { LocaleModule } from '@farris/ui-locale';
export class NotifyModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: NotifyModule,
            providers: [NotifyService]
        };
    }
}
NotifyModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FarrisCommonModule,
                    LocaleModule
                ],
                declarations: [NotifyComponent, NotifyContainerComponent],
                providers: [
                    NotifyService
                ],
                exports: [
                    NotifyComponent,
                    NotifyContainerComponent
                ],
                entryComponents: [NotifyContainerComponent]
            },] }
];
