// @farris/mobile-devkit/esm2015/lib/expression/listener/ui_state_change_listener.js
import { ChangeListener } from './change_listener';
import { Expression } from '../expression/index';
import { NAMESPACE } from '../core/index';
/**
 * 监听UIState变更
 */
class UIStateChangeListener extends ChangeListener {
    /**
     * 构造函数
     */
    constructor(viewModelContext) {
        super();
        this.viewModelContext = viewModelContext;
        this.frameId = this.viewModelContext.id;
        this.uiState = this.viewModelContext.uiState;
        this.namespace = this.viewModelContext.injector.get(NAMESPACE, null);
        this.registerEvent();
    }
    buildEventPath(change) {
        return null;
    }
    registerEvent() {
        if (this.uiState && this.uiState.changes) {
            this.uiState.changes.subscribe((change) => {
                const modification = {
                    ns: this.namespace,
                    path: [change.field],
                    type: Expression.EventType.ValueChanged,
                    value: change.value,
                    source: Expression.EventSource.State,
                    frameId: this.frameId
                };
                // console.log("UIStateChangeListener", modification);
                this.subject.next(modification);
            });
        }
    }
}
export { UIStateChangeListener };
