// @gsp-svc/file-viewer/esm5/lib/pdf-viewer/entitires/pdfviewoptions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var PdfViewOptions = /** @class */ (function () {
    function PdfViewOptions() {
        this.showHeader = true;
        this.showPaging = true;
        this.showThumbnails = true;
        this.showDownload = true;
        this.showSearch = true;
        this.showZoom = true;
        this.showPrint = true;
        this.showSecondaryToolBar = true;
    }
    return PdfViewOptions;
}());
export { PdfViewOptions };
if (false) {
    /** @type {?} */
    PdfViewOptions.prototype.showHeader;
    /** @type {?} */
    PdfViewOptions.prototype.showPaging;
    /** @type {?} */
    PdfViewOptions.prototype.showThumbnails;
    /** @type {?} */
    PdfViewOptions.prototype.showDownload;
    /** @type {?} */
    PdfViewOptions.prototype.showSearch;
    /** @type {?} */
    PdfViewOptions.prototype.showZoom;
    /** @type {?} */
    PdfViewOptions.prototype.showPrint;
    /** @type {?} */
    PdfViewOptions.prototype.showSecondaryToolBar;
}
