// @farris/ide-property-panel/esm2015/lib/property-panel/controls/combolist.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output } from '@angular/core';
export class CombolistPropertyComponent {
    constructor() {
        this.valueChanged = new EventEmitter();
        /**
         * 下拉框的枚举项
         */
        this.selectOptions = [];
        /**
         * 是否可任意输入
         */
        this.nosearch = false;
        /**
         * 是否多选
         */
        this.multiSelect = false;
    }
    /**
     * @return {?}
     */
    get elementValue() {
        return this._elementValue;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set elementValue(value) {
        this._elementValue = value;
        // 多选控件使用数组格式的数据
        if (this.elementConfig.multiSelectDataType === 'array') {
            this._elementValue = this._elementValue ? this._elementValue.toString() : '';
        }
    }
    /**
     * @return {?}
     */
    get elementConfig() {
        return this._elementConfig;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set elementConfig(value) {
        this._elementConfig = value;
        if (!this.elementConfig) {
            return;
        }
        this.editType = this.elementConfig.propertyType;
        if (this.editType === 'boolean') {
            this.selectOptions = [{ key: true, value: 'true' }, { key: false, value: 'false' }];
            this.editType = 'select';
        }
        else {
            this.selectOptions = this.elementConfig.iterator || [];
        }
        switch (this.editType) {
            case 'select': {
                // 不可编辑下拉框
                this.editable = false;
                break;
            }
            case 'editableSelect': {
                // 可编辑下拉框
                this.editable = true;
                this.nosearch = true;
                break;
            }
            case 'multiSelect': {
                // 不可编辑的下拉多选框
                this.editable = false;
                this.multiSelect = true;
                // 多选控件使用数组格式的数据
                if (this.elementConfig.multiSelectDataType === 'array') {
                    this._elementValue = this._elementValue ? this._elementValue.toString() : '';
                }
                break;
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @param {?} $event
     * @return {?}
     */
    changeValue($event) {
        /** @type {?} */
        let value = this._elementValue;
        // 多选控件返回数组格式的数据
        if (this.editType === 'multiSelect' && this.elementConfig.multiSelectDataType === 'array') {
            value = this._elementValue ? this._elementValue.split(',') : '';
        }
        this.valueChanged.next({ elementValue: value });
    }
}
CombolistPropertyComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-combolist-prop',
                template: `
      <!--下拉列表-->
      <div class="farris-input-wrap">
        <farris-combo-list
          [(ngModel)]="elementValue"
          [idField]="'key'"
          [textField]="'value'"
          [data]="selectOptions"
          [readonly]="elementConfig.readonly"
          [editable]="editable"
          [multiSelect]="multiSelect"
          [enableClear]="false"
          [enableCancelSelected]="false"
          [nosearch]="nosearch"
          (ngModelChange)="changeValue($event)">
        </farris-combo-list>
      </div>
    `
            }] }
];
CombolistPropertyComponent.propDecorators = {
    valueChanged: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CombolistPropertyComponent.prototype.valueChanged;
    /**
     * @type {?}
     * @private
     */
    CombolistPropertyComponent.prototype._elementConfig;
    /**
     * @type {?}
     * @private
     */
    CombolistPropertyComponent.prototype._elementValue;
    /**
     * 编辑类型
     * @type {?}
     */
    CombolistPropertyComponent.prototype.editType;
    /**
     * 下拉框的枚举项
     * @type {?}
     */
    CombolistPropertyComponent.prototype.selectOptions;
    /**
     * 是否可编辑
     * @type {?}
     */
    CombolistPropertyComponent.prototype.editable;
    /**
     * 是否可任意输入
     * @type {?}
     */
    CombolistPropertyComponent.prototype.nosearch;
    /**
     * 是否多选
     * @type {?}
     */
    CombolistPropertyComponent.prototype.multiSelect;
}
