// @farris/ide-property-panel/esm2015/lib/property-panel/controls/date.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, EventEmitter, Output } from '@angular/core';
export class DatePropertyComponent {
    /**
     * @param {?} cd
     */
    constructor(cd) {
        this.cd = cd;
        this.showTime = false;
        this.dateFormat = 'yyyy-MM-dd';
        this.valueChanged = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get elementConfig() {
        return this._elementConfig;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set elementConfig(value) {
        this._elementConfig = value;
        if (!this.elementConfig) {
            return;
        }
        this.maxDate = value.max;
        this.minDate = value.min;
        this.showTime = this.elementConfig.propertyType === 'datetime';
        this.dateFormat = this.elementConfig.propertyType === 'datetime' ? 'yyyy-MM-dd HH:mm:ss' : 'yyyy-MM-dd';
    }
    /**
     * @return {?}
     */
    get elementValue() {
        return this._elementValue;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set elementValue(value) {
        this._elementValue = value;
    }
    // editType: string;
    /**
     * @return {?}
     */
    ngOnInit() {
        // this.editType = this.elementConfig && this.elementConfig.propertyType;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    changeValue($event) {
        /** @type {?} */
        const value = $event.formatted;
        this.valueChanged.next({ $event, elementValue: value });
    }
}
DatePropertyComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-date-prop',
                template: `
        <farris-datepicker [(ngModel)]="elementValue" [editable]="false" [showTime]="showTime"
          [dateFormat]="dateFormat" [readonly]="elementConfig.readonly"
          (valueChange)="changeValue($event)" [returnType]="'string'" [maxDate]="maxDate"
          [minDate]="minDate">
        </farris-datepicker>

    `
            }] }
];
/** @nocollapse */
DatePropertyComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
DatePropertyComponent.propDecorators = {
    valueChanged: [{ type: Output }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    DatePropertyComponent.prototype._elementConfig;
    /**
     * @type {?}
     * @private
     */
    DatePropertyComponent.prototype._elementValue;
    /** @type {?} */
    DatePropertyComponent.prototype.maxDate;
    /** @type {?} */
    DatePropertyComponent.prototype.minDate;
    /** @type {?} */
    DatePropertyComponent.prototype.showTime;
    /** @type {?} */
    DatePropertyComponent.prototype.dateFormat;
    /** @type {?} */
    DatePropertyComponent.prototype.valueChanged;
    /**
     * @type {?}
     * @private
     */
    DatePropertyComponent.prototype.cd;
}
