// @farris/ide-property-panel/esm2015/lib/property-panel/controls/number.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output } from '@angular/core';
export class NumberPropertyComponent {
    constructor() {
        this.valueChanged = new EventEmitter();
        /**
         * 数字控件精度
         */
        this.precision = 2;
    }
    /**
     * @return {?}
     */
    get elementConfig() {
        return this._elementConfig;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set elementConfig(value) {
        this._elementConfig = value;
        this.precision = this._elementConfig.decimals === undefined ? 2 : this._elementConfig.decimals;
    }
    /**
     * @return {?}
     */
    get elementValue() {
        return this._elementValue;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set elementValue(value) {
        this._elementValue = value;
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @param {?} $event
     * @return {?}
     */
    changeValue($event) {
        this.valueChanged.next({ $event, elementValue: this.elementValue });
    }
}
NumberPropertyComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-number-prop',
                template: `
    <div class="farris-input-wrap">
        <farris-number-spinner [(ngModel)]="elementValue" [readonly]="elementConfig.readonly" [precision]="precision"
            [autoDecimal]="true" [min]="elementConfig.min" [max]="elementConfig.max" [canNull]="true"
            [bigNumber]="elementConfig.isBigNumber" (ngModelChange)="changeValue($event)">
        </farris-number-spinner>
    </div>
    `
            }] }
];
NumberPropertyComponent.propDecorators = {
    valueChanged: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    NumberPropertyComponent.prototype.valueChanged;
    /**
     * 数字控件精度
     * @type {?}
     */
    NumberPropertyComponent.prototype.precision;
    /**
     * @type {?}
     * @private
     */
    NumberPropertyComponent.prototype._elementConfig;
    /**
     * @type {?}
     * @private
     */
    NumberPropertyComponent.prototype._elementValue;
}
