// @farris/ide-property-panel/esm2015/lib/property-panel/controls/textbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output } from '@angular/core';
export class TextboxPropertyComponent {
    constructor() {
        this.valueChanged = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get elementConfig() {
        return this._elementConfig;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set elementConfig(value) {
        this._elementConfig = value;
        if (this.elementConfig.readonly) {
            this.convertModalShowValue(this._elementValue);
        }
    }
    /**
     * @return {?}
     */
    get elementValue() {
        return this._elementValue;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set elementValue(value) {
        this._elementValue = value;
        if (this.elementConfig.readonly) {
            this.convertModalShowValue(this._elementValue);
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    changeValue($event) {
        // 文本控件限制输入字符
        if (this.elementValue && !this.elementConfig.readonly &&
            this.elementConfig.notAllowedChars && this.elementConfig.notAllowedChars.length) {
            this.elementConfig.notAllowedChars.forEach((/**
             * @param {?} c
             * @return {?}
             */
            (c) => {
                if (typeof (c) === 'object') {
                    this.elementValue = this.elementValue.replace(c, '');
                }
                else if (typeof (c) === 'string') {
                    this.elementValue = this.elementValue.replace(new RegExp(c, 'g'), '');
                }
            }));
        }
        this.valueChanged.next({ $event, elementValue: this.elementValue });
    }
    /**
     * @param {?} value
     * @return {?}
     */
    restValueAfterValidRestrict(value) {
        this.elementValue = value || '';
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    convertModalShowValue(value) {
        if (this.elementConfig.converter) {
            this.elementShowValue = this.elementConfig.converter.convertTo(value);
            return;
        }
        if (value && value instanceof Object) {
            this.elementShowValue = JSON.stringify(value);
            return;
        }
        this.elementShowValue = value;
    }
}
TextboxPropertyComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-textbox-prop',
                template: `
    <div class="farris-input-wrap">
        <input type="input" class="form-control form-control-sm" [(ngModel)]="elementValue" (blur)="changeValue($event)"
            *ngIf="!elementConfig.readonly">
        <input type="input" class="form-control form-control-sm" [(ngModel)]="elementShowValue" [readonly]="'readonly'"
            *ngIf="elementConfig.readonly">
    </div>
    `
            }] }
];
TextboxPropertyComponent.propDecorators = {
    valueChanged: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    TextboxPropertyComponent.prototype.valueChanged;
    /**
     * @type {?}
     * @private
     */
    TextboxPropertyComponent.prototype._elementConfig;
    /**
     * @type {?}
     * @private
     */
    TextboxPropertyComponent.prototype._elementValue;
    /**
     * 输入框中显示的值
     * @type {?}
     */
    TextboxPropertyComponent.prototype.elementShowValue;
}
