// @farris/ide-property-panel/esm2015/lib/property-panel/controls/unity.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { PropertyEntity } from '../entity/property-entity';
export class UniformPropertyComponent {
    constructor() {
        this.valueChanged = new EventEmitter();
    }
    /**
     * @return {?}
     */
    initImportedOriginalData() {
        /** @type {?} */
        const editorParams = this.elementConfig.editorParams;
        /** @type {?} */
        const constType = editorParams.constType;
        /** @type {?} */
        const valueType = this.getValueType();
        this.importedOriginalData = {
            controlName: editorParams.controlName,
            propertyName: this.elementConfig.propertyID,
            propertyType: constType,
            propertyValue: {
                type: valueType,
                value: this.elementValue
            },
            editorOptions: editorParams.editorOptions
        };
    }
    /**
     * 通过当前配置信息和值信息推断当前值的类型
     * @return {?} 值类型
     */
    getValueType() {
        /** @type {?} */
        const propertyValue = this.elementValue;
        /** @type {?} */
        const editorParams = this.elementConfig.editorParams;
        /** @type {?} */
        const constType = editorParams.constType;
        /** @type {?} */
        const types = editorParams.editorOptions.types;
        // 当前值为对象时
        if (propertyValue && typeof propertyValue === "object") {
            // 判断是否为变量类型
            if (propertyValue.hasOwnProperty("type") && (typeof propertyValue.type === "string" && propertyValue.type.toLowerCase() === "variable")) {
                if (types.includes("variable"))
                    return "variable";
            }
        }
        // 当前值为字符串时
        if (typeof propertyValue === "string") {
            // 判断是否为自定义类型（通过是否含有“viewModel”子串硬推断）
            if (types.includes("custom") && (propertyValue.includes("viewModel") || !types.includes("const") || constType !== "string")) {
                return "custom";
            }
            // 判断是否为字符串常量
            if (types.includes("const") && constType === "string") {
                return "const";
            }
        }
        // 当前值为数字时
        if (typeof propertyValue === "number") {
            if (types.includes("const") && constType === "number") {
                return "const";
            }
        }
        // 默认为const类型（为枚举类型兜底）
        if (types.includes("const")) {
            return "const";
        }
        // 如果都找不到则任意返回一个值
        return types[0];
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    handleValueChanged($event) {
        /** @type {?} */
        let newValue = $event.propertyValue.value;
        if ($event.propertyValue.type === "variable" && typeof newValue === "object") {
            newValue.type = "Variable";
            if (newValue.name) {
                delete newValue.name;
            }
            if ($event["isNewVariable"]) {
                newValue["isNewVariable"] = true;
            }
        }
        this.valueChanged.next({ $event, elementValue: newValue });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.initImportedOriginalData();
    }
}
UniformPropertyComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-unity-prop',
                template: `
      <div class="farris-input-wrap">
        <farris-propertyEditor
          [importedOriginalData]="importedOriginalData"
          (propertyEditorValueChanged)="handleValueChanged($event)">
        </farris-propertyEditor>
      </div>
    `
            }] }
];
UniformPropertyComponent.propDecorators = {
    elementConfig: [{ type: Input }],
    elementValue: [{ type: Input }],
    valueChanged: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    UniformPropertyComponent.prototype.elementConfig;
    /** @type {?} */
    UniformPropertyComponent.prototype.elementValue;
    /** @type {?} */
    UniformPropertyComponent.prototype.valueChanged;
    /** @type {?} */
    UniformPropertyComponent.prototype.importedOriginalData;
}
