// @gsp-svc/file-viewer/esm5/lib/pdf-viewer/pdf-viewer/pdf-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { PdfViewOptions } from '../entitires/pdfviewoptions';
var PdfViewerComponent = /** @class */ (function () {
    function PdfViewerComponent(sanitizer) {
        this.sanitizer = sanitizer;
        this.pdfUrl = "";
        this.baseUrl = "/platform/common/web/pdfjs/viewer.html?file=";
    }
    /**
     * @return {?}
     */
    PdfViewerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.viewUrl = this.sanitizer.bypassSecurityTrustResourceUrl(this.getViewUrl(this.baseUrl + encodeURIComponent(this.pdfUrl)));
    };
    /**
     * @param {?} url
     * @return {?}
     */
    PdfViewerComponent.prototype.pdfChange = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.viewUrl = this.sanitizer.bypassSecurityTrustResourceUrl(this.getViewUrl(this.baseUrl + encodeURIComponent(url)));
    };
    /**
     * @param {?} url
     * @return {?}
     */
    PdfViewerComponent.prototype.getViewUrl = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        if (!this.viewOptions.showPaging)
            url += "&showpaging=false";
        if (!this.viewOptions.showHeader)
            url += "&showheader=false";
        if (!this.viewOptions.showThumbnails)
            url += "&showthumbnails=false";
        if (!this.viewOptions.showDownload)
            url += "&showdownload=false";
        if (!this.viewOptions.showSearch)
            url += "&showsearch=false";
        if (!this.viewOptions.showPrint)
            url += "&showprint=false";
        if (!this.viewOptions.showZoom)
            url += "&showzoom=false";
        if (!this.viewOptions.showSecondaryToolBar)
            url += "&showsecondarytoolbar=false";
        return url;
    };
    PdfViewerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-pdf-viewer',
                    template: "<iframe id=\"viewframe\" [src]=\"viewUrl\" height=\"100%\" width=\"100%\"></iframe>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    PdfViewerComponent.ctorParameters = function () { return [
        { type: DomSanitizer }
    ]; };
    PdfViewerComponent.propDecorators = {
        pdfUrl: [{ type: Input }],
        viewOptions: [{ type: Input }]
    };
    return PdfViewerComponent;
}());
export { PdfViewerComponent };
if (false) {
    /** @type {?} */
    PdfViewerComponent.prototype.pdfUrl;
    /** @type {?} */
    PdfViewerComponent.prototype.viewOptions;
    /** @type {?} */
    PdfViewerComponent.prototype.baseUrl;
    /** @type {?} */
    PdfViewerComponent.prototype.viewUrl;
    /**
     * @type {?}
     * @private
     */
    PdfViewerComponent.prototype.sanitizer;
}
