// @ecp-caf/caf-common/esm2015/lib/services/tabservice/tab-entity.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { WindowVariableService } from '../window-service/window-variable.service';
import * as i0 from "@angular/core";
/**
 * @record
 */
function TabEntity() { }
if (false) {
    /** @type {?} */
    TabEntity.prototype.tabId;
    /** @type {?|undefined} */
    TabEntity.prototype.code;
    /** @type {?|undefined} */
    TabEntity.prototype.name;
    /** @type {?} */
    TabEntity.prototype.appType;
    /** @type {?} */
    TabEntity.prototype.funcId;
    /** @type {?} */
    TabEntity.prototype.appId;
    /** @type {?} */
    TabEntity.prototype.appEntrance;
}
export class TabEntityService {
    constructor() {
        this.nameSpace = 'gsp-rtf-tabarr-service';
        this.tabarrayKey = 'tab-arry';
        this.windowService = new WindowVariableService();
        this.topWin = this.windowService.getTopWindow();
        this.topWin[this.nameSpace] = this.topWin[this.nameSpace] || {};
    }
    /**
     * @param {?} tabs
     * @return {?}
     */
    setTabs(tabs) {
        if (!tabs || tabs.length === 0) {
            return;
        }
        /** @type {?} */
        const copy = this.deepClone(tabs);
        this.topWin[this.nameSpace][this.tabarrayKey] = copy;
    }
    /**
     * @return {?}
     */
    getTabs() {
        /** @type {?} */
        const tabs = this.topWin[this.nameSpace][this.tabarrayKey] || [];
        if (!tabs || tabs.length === 0) {
            return tabs;
        }
        return this.deepClone(tabs);
    }
    /**
     * @param {?} tab
     * @return {?}
     */
    setActiveTab(tab) {
        /** @type {?} */
        const tabs = this.getTabs();
        /** @type {?} */
        const activeOld = tabs.find((/**
         * @param {?} t
         * @return {?}
         */
        t => t.active));
        if (!tab) {
            if (!!activeOld) {
                activeOld.active = false;
            }
            return;
        }
        if (tabs.length > 0) {
            /** @type {?} */
            let activeNew = tabs.find((/**
             * @param {?} t
             * @return {?}
             */
            t => t.id === tab.id));
            if (!activeNew) {
                activeNew = tab;
                if (!!activeOld) {
                    activeOld.active = false;
                }
                activeNew.active = true;
                tabs.push(activeNew);
            }
            else {
                if (!!activeOld) {
                    if (activeOld === activeNew) {
                        return;
                    }
                    else {
                        activeOld.active = false;
                        activeNew.active = true;
                    }
                }
                else {
                    activeNew.active = true;
                }
            }
            this.setTabs(tabs);
        }
    }
    /**
     * @return {?}
     */
    getActiveTab() {
        /** @type {?} */
        const tabs = this.getTabs();
        /** @type {?} */
        const activeTab = tabs && tabs.find((/**
         * @param {?} t
         * @return {?}
         */
        t => t.active));
        /** @type {?} */
        const res = activeTab && {
            tabId: activeTab.id,
            appType: activeTab.appType,
            funcId: activeTab.funcId,
            appId: activeTab.appId,
            appEntrance: activeTab.appEntrance,
            name: activeTab.FuncName
        };
        return res;
    }
    /**
     * @private
     * @param {?} target
     * @return {?}
     */
    deepClone(target) {
        /** @type {?} */
        let result;
        if (typeof target === 'object') {
            if (Array.isArray(target)) {
                result = [];
                for (let i in target) {
                    result.push(this.deepClone(target[i]));
                }
            }
            else if (target === null) {
                result = null;
            }
            else if (target.constructor === RegExp) {
                result = target;
            }
            else {
                result = {};
                for (let i in target) {
                    result[i] = this.deepClone(target[i]);
                }
            }
        }
        else {
            result = target;
        }
        return result;
    }
}
TabEntityService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TabEntityService.ctorParameters = () => [];
/** @nocollapse */ TabEntityService.ngInjectableDef = i0.defineInjectable({ factory: function TabEntityService_Factory() { return new TabEntityService(); }, token: TabEntityService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    TabEntityService.prototype.nameSpace;
    /**
     * @type {?}
     * @private
     */
    TabEntityService.prototype.tabarrayKey;
    /**
     * @type {?}
     * @private
     */
    TabEntityService.prototype.windowService;
    /**
     * @type {?}
     * @private
     */
    TabEntityService.prototype.topWin;
}
