// @farris/ui-tabs/esm5/lib/tab/tab.transclude.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ViewContainerRef, TemplateRef, KeyValueDiffers } from '@angular/core';
var TabTranscludeDirective = /** @class */ (function () {
    function TabTranscludeDirective(viewRef, _differs) {
        this._differs = _differs;
        // setTimeout(() => {
        this.viewRef = viewRef;
        // }, 5000);
        // this.cd.detectChanges();
    }
    Object.defineProperty(TabTranscludeDirective.prototype, "trancludeTempl", {
        get: /**
         * @return {?}
         */
        function () {
            return this._trancludeTempl;
        },
        set: /**
         * @param {?} templateRef
         * @return {?}
         */
        function (templateRef) {
            this._trancludeTempl = templateRef;
            if (templateRef) {
                this.viewRef.createEmbeddedView(templateRef);
            }
            // if (!this._differ && templateRef) {
            //     this._differ = this._differs.find(templateRef).create();
            // }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TabTranscludeDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        // this.setIntervalTemp = setInterval(() => {
        //     if (this._trancludeTempl && this.viewRef) {
        //         console.log('sa');
        //         // this.viewRef = this.viewRef;
        //         clearInterval(this.setIntervalTemp);
        //     }
        // }, 0);
    };
    /**
     * @return {?}
     */
    TabTranscludeDirective.prototype.ngDoCheck = /**
     * @return {?}
     */
    function () {
        // const CHANGE = this._differ.diff(this._trancludeTempl);
        // if (CHANGE) {
        //     this.viewRef.createEmbeddedView(this._trancludeTempl);
        // }
    };
    TabTranscludeDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[tabTransclude]',
                },] }
    ];
    /** @nocollapse */
    TabTranscludeDirective.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: KeyValueDiffers }
    ]; };
    TabTranscludeDirective.propDecorators = {
        trancludeTempl: [{ type: Input, args: ['tabTransclude',] }]
    };
    return TabTranscludeDirective;
}());
export { TabTranscludeDirective };
if (false) {
    /** @type {?} */
    TabTranscludeDirective.prototype._trancludeTempl;
    /** @type {?} */
    TabTranscludeDirective.prototype.setIntervalTemp;
    /**
     * @type {?}
     * @private
     */
    TabTranscludeDirective.prototype._differ;
    /** @type {?} */
    TabTranscludeDirective.prototype.viewRef;
    /**
     * @type {?}
     * @private
     */
    TabTranscludeDirective.prototype._differs;
}
