// @qdp/spread/esm5/lib/service/render/render.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, Optional } from '@angular/core';
import { SchemaManagerService, LcpService } from '@qdp/common';
import { CacheService } from '@ecp-caf/caf-common';
import { LoadingService } from '@farris/ui-loading';
import { NotifyService } from '@farris/ui-notify';
import { DataRenderExtendService } from '@qdp/common';
import { RtfServices } from '@qdp/common';
import { LocalizeService } from '@qdp/localize';
import { MessagerService } from '@farris/ui-messager';
import { FormErrorService } from '@farris/command-services';
import { CloudprintService } from '@gsp-svc/cloudprint';
import { ServicesUtils } from '../../util/services.utils';
import { CrosstabRender } from '../../engine/crosstab-render';
import { ListTreeRender } from '../../engine/list-render';
var RenderService = /** @class */ (function () {
    function RenderService(schemaManager, lcpService, cache, loadService, msgService, dataRenderExtendService, injector) {
        this.schemaManager = schemaManager;
        this.lcpService = lcpService;
        this.cache = cache;
        this.loadService = loadService;
        this.msgService = msgService;
        this.dataRenderExtendService = dataRenderExtendService;
        this.injector = injector;
        this.serviceCache = {};
        this.columnCount = 0;
        this.operations = {};
        this.funcInfoCache = {
            funcIds: [],
            firstLoadCache: []
        };
        /** @type {?} */
        var self = this;
        if (this.injector) {
            this.localizeService = this.injector.get(LocalizeService);
            this.formErrorService = this.injector.get(FormErrorService);
            this.cloudprintService = this.injector.get(CloudprintService);
            this.notifyService = this.injector.get(NotifyService);
            this.setService();
            self.schemaManager.operationChange.subscribe((/**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                if (value) {
                    self.operations = value;
                }
            }));
            self.setSecurityLevel();
        }
    }
    /**
     * @private
     * @return {?}
     */
    RenderService.prototype.setService = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var tabId = RtfServices.getTabId() || 'default';
        this.serviceCache[tabId] = {
            'formErrorService': this.formErrorService,
            'cloudprintService': this.cloudprintService,
            'schemaManagerService': this.schemaManager,
            'lcpService': this.lcpService,
            'cacheService': this.cache,
            'loadingService': this.loadService,
            'msgService': this.msgService,
            'localizeService': this.localizeService,
            'notifyService': this.notifyService
        };
    };
    /**
     * @private
     * @return {?}
     */
    RenderService.prototype.getService = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var tabId = RtfServices.getTabId() || 'default';
        return this.serviceCache[tabId];
    };
    /**
     * @private
     * @return {?}
     */
    RenderService.prototype.setSecurityLevel = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.lcpService.variablesChange.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data && data.hasOwnProperty('securityLevel') && data['securityLevel']) {
                _this.cache.set(RtfServices.getTabId() + 'securityLevel', data['securityLevel']);
            }
        }));
    };
    /**
     * @private
     * @return {?}
     */
    RenderService.prototype.getSecurityLevel = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.cache.get(RtfServices.getTabId() + 'securityLevel')) {
            return this.cache.get(RtfServices.getTabId() + 'securityLevel');
        }
        else {
            return '';
        }
    };
    /**
     * @param {?} option
     * @return {?}
     */
    RenderService.prototype.render = /**
     * @param {?} option
     * @return {?}
     */
    function (option) {
        if (!option.loading) {
            option.loading = this.loadService.show({ container: 'body' });
        }
        this.spread = option.spread ? option.spread : null;
        this.renderObject = this.getRender(option);
        this.renderObject.render(option);
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    RenderService.prototype.getRender = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        this.setService();
        ServicesUtils.setServices('dataRenderExtendService' + obj.queryId, this.dataRenderExtendService);
        /** @type {?} */
        var render;
        switch (obj.controlType) {
            case 'crosstab':
                render = new CrosstabRender(obj.queryId, obj.controlType, obj.jointsearchInfoList, this.funcInfoCache, this.getService());
                break;
            default:
                render = new ListTreeRender(obj.queryId, obj.controlType, obj.jointsearchInfoList, this.funcInfoCache, this.getService());
                break;
        }
        render.isSchemaSelectChanged = obj.isSchemaSelectChanged ? obj.isSchemaSelectChanged : false;
        return render;
    };
    /**
     * @param {?} renderOption
     * @return {?}
     */
    RenderService.prototype.exportData = /**
     * @param {?} renderOption
     * @return {?}
     */
    function (renderOption) {
        var _this = this;
        // exportData(queryId: any, schemaId: any, controlType: any, pageIndex?: number, qoManagerCode?: any, queryRelativeUrl?: any, filterCondition?: any, groupType?: any, voId?: any, totalCount?: any, exportType?: any) {
        /** @type {?} */
        var loading = this.loadService.show({ container: 'body' });
        this.renderObject.utils.get('schema').getSchemaInfo({ schemaId: renderOption.schemaValue, queryId: renderOption.queryId, queryRelativeUrl: renderOption.queryRelativeUrl, groupType: renderOption.groupType }).subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            renderOption.schemaId = renderOption.schemaId ? renderOption.schemaId : (value && value.id ? value.id : '1');
            /** @type {?} */
            var schemaName = value && value.schemaValue && value.schemaValue.titleOption && value.schemaValue.titleOption.title
                ? value.schemaValue.titleOption.title
                : value && value.schemaName
                    ? value.schemaName
                    : '';
            if (RtfServices.getFuncName()) {
                schemaName = RtfServices.getFuncName() + '-' + schemaName;
            }
            renderOption.excelExportName = schemaName;
            /** @type {?} */
            var params = _this.renderObject.utils.get('parameter').createParameters(renderOption);
            params['exportType'] = renderOption.exportType ? renderOption.exportType : '';
            /** @type {?} */
            var isCustomShema = false;
            /** @type {?} */
            var customSchema = null;
            if (params['schemaId'] === '1') {
                if (ServicesUtils.getServices(RtfServices.getTabId(renderOption.queryId) + 'customScheme')) {
                    isCustomShema = true;
                    params.customSchema = ServicesUtils.getServices(RtfServices.getTabId(renderOption.queryId) + 'customScheme');
                    customSchema = JSON.parse(params.customSchema);
                    if (!params.excelExportName) {
                        params.excelExportName = customSchema && customSchema.titleOption && customSchema.titleOption.title
                            ? customSchema.titleOption.title : 'export';
                    }
                }
                else {
                    _this.msgService.warning(_this.localizeService.getValue('spread.message.export'));
                    loading.close();
                    return;
                }
            }
            if (_this.operations['QdpQueryExport'] === false) {
                _this.msgService.warning(_this.localizeService.getValue('spread.message.validate.qdpQueryExport'));
                loading.close();
                return;
            }
            try {
                if (renderOption.controlType === 'treelist' && _this.spread) {
                    /** @type {?} */
                    var expandNodes = '';
                    if (value && value.schemaValue && value.schemaValue.otherOption && value.schemaValue.otherOption.treeInfoType !== 2) {
                        /** @type {?} */
                        var pathField = value.schemaValue.otherOption.treeInfoType === 0 ? value.schemaValue.otherOption.pathField : value.schemaValue.otherOption.idField;
                        if (pathField) {
                            for (var i = 0; i < _this.spread.getActiveSheet().getRowCount(); i++) {
                                if (_this.spread.getActiveSheet().rowOutlines.isCollapsed(i)) {
                                    expandNodes += ',' + _this.spread.getActiveSheet().getDataItem(i)[pathField];
                                }
                            }
                            params['expandNodes'] = expandNodes.substring(1);
                        }
                    }
                }
            }
            catch (e) {
            }
            _this.columnCount = 0;
            if (!isCustomShema) {
                _this.calculationColumnCount(value.schemaValue.columnOption.colList);
            }
            else {
                if (customSchema && customSchema.columnOption && customSchema.columnOption.colList) {
                    _this.calculationColumnCount(customSchema.columnOption.colList);
                }
            }
            _this.lcpService.loading = loading;
            if (_this.getSecurityLevel()) {
                params.excelExportName += '_' + _this.getSecurityLevel();
            }
            if ((_this.columnCount * renderOption.totalCount) < 100000) {
                _this.lcpService.exportData(params, renderOption.queryRelativeUrl);
                loading.close();
            }
            else {
                _this.lcpService.aysnExportData(params, renderOption.queryRelativeUrl);
                loading.close();
                _this.msgService.warning(_this.localizeService.getValue('spread.message.aysnExport'));
            }
        }));
    };
    /**
     * @private
     * @param {?} colList
     * @return {?}
     */
    RenderService.prototype.calculationColumnCount = /**
     * @private
     * @param {?} colList
     * @return {?}
     */
    function (colList) {
        var _this = this;
        try {
            colList.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.childList && item.childList.length) {
                    _this.calculationColumnCount(item.childList);
                }
                else {
                    _this.columnCount += 1;
                }
            }));
        }
        catch (e) {
            this.columnCount = 0;
        }
    };
    /**
     * @param {?} queryId
     * @param {?} schemaId
     * @param {?} controlType
     * @param {?=} qoManagerCode
     * @param {?=} queryRelativeUrl
     * @param {?=} filterCondition
     * @param {?=} groupType
     * @param {?=} voId
     * @return {?}
     */
    RenderService.prototype.clearCache = /**
     * @param {?} queryId
     * @param {?} schemaId
     * @param {?} controlType
     * @param {?=} qoManagerCode
     * @param {?=} queryRelativeUrl
     * @param {?=} filterCondition
     * @param {?=} groupType
     * @param {?=} voId
     * @return {?}
     */
    function (queryId, schemaId, controlType, qoManagerCode, queryRelativeUrl, filterCondition, groupType, voId) {
        var _this = this;
        this.renderObject.utils.get('schema').getSchemaInfo({ schemaId: schemaId, queryId: queryId, queryRelativeUrl: queryRelativeUrl, groupType: groupType }).subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            schemaId = schemaId ? schemaId : (value && value.id ? value.id : '1');
            /** @type {?} */
            var schemaName = value && value.schemaValue && value.schemaValue.titleOption && value.schemaValue.titleOption.title
                ? value.schemaValue.titleOption.title
                : value && value.schemaName
                    ? value.schemaName
                    : '';
            /** @type {?} */
            var params = _this.renderObject.utils.get('parameter').createParameters({
                queryId: queryId,
                controlType: controlType,
                schemaId: schemaId,
                qoManagerCode: qoManagerCode,
                pageIndex: 0,
                pageSize: 0,
                excelExportName: schemaName,
                filterCondition: filterCondition,
                voId: voId
            });
            if (ServicesUtils.getServices(RtfServices.getTabId(queryId) + 'customScheme')) {
                params.customSchema = ServicesUtils.getServices(RtfServices.getTabId(queryId) + 'customScheme');
            }
            _this.lcpService.clearCache(params).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                if (data) {
                    _this.notifyService['success']((/** @type {?} */ ({
                        title: _this.localizeService.getValue('ideCmp.schemaManager.message.info.title'), msg: _this.localizeService.getValue('spread.message.clearCache'), timeout: 3000
                    })));
                }
            }));
        }));
    };
    /**
     * @param {?} renderOption
     * @return {?}
     */
    RenderService.prototype.pdfExportEvent = /**
     * @param {?} renderOption
     * @return {?}
     */
    function (renderOption) {
        var _this = this;
        // pdfExportEvent(queryId: any, schemaId: any, controlType: any, qoManagerCode?: any, printIntegration?: any, queryRelativeUrl?: any, filterCondition?: any, groupType?: any, pageIndex?: any, voId?: any) {
        /** @type {?} */
        var loading = this.loadService.show({ container: 'body' });
        try {
            this.renderObject.utils.get('schema').getSchemaInfo({ schemaId: renderOption.schemaValue, queryId: renderOption.queryId, queryRelativeUrl: renderOption.queryRelativeUrl, groupType: renderOption.groupType }).subscribe((/**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                renderOption.schemaId = renderOption.schemaId ? renderOption.schemaId : (value && value.id ? value.id : '1');
                /** @type {?} */
                var schemaName = value && value.schemaValue && value.schemaValue.titleOption && value.schemaValue.titleOption.title
                    ? value.schemaValue.titleOption.title
                    : value && value.schemaName
                        ? value.schemaName
                        : '';
                if (RtfServices.getFuncName()) {
                    schemaName = RtfServices.getFuncName() + '-' + schemaName;
                }
                renderOption.excelExportName = schemaName;
                /** @type {?} */
                var params = _this.renderObject.utils.get('parameter').createParameters(renderOption);
                if (params['schemaId'] === '1') {
                    if (ServicesUtils.getServices(RtfServices.getTabId(renderOption.queryId) + 'customScheme')) {
                        params.customSchema = ServicesUtils.getServices(RtfServices.getTabId(renderOption.queryId) + 'customScheme');
                        /** @type {?} */
                        var customSchema = JSON.parse(params.customSchema);
                        if (!params.excelExportName) {
                            params.excelExportName = customSchema && customSchema.titleOption && customSchema.titleOption.title
                                ? customSchema.titleOption.title : 'export';
                        }
                    }
                    else {
                        _this.msgService.warning(_this.localizeService.getValue('spread.message.print'));
                        loading.close();
                        return;
                    }
                }
                if (_this.operations['QdpQueryPrint'] === false) {
                    _this.msgService.warning(_this.localizeService.getValue('spread.message.validate.qdpQueryPrint'));
                    loading.close();
                    return;
                }
                if (_this.getSecurityLevel()) {
                    params.excelExportName += '_' + _this.getSecurityLevel();
                }
                _this.lcpService.pdfExportEvent(params, renderOption.queryRelativeUrl);
                loading.close();
            }));
        }
        catch (e) {
            loading.close();
            // console.log(e);
        }
    };
    /**
     * @param {?} renderOption
     * @return {?}
     */
    RenderService.prototype.printPreView = /**
     * @param {?} renderOption
     * @return {?}
     */
    function (renderOption) {
        var _this = this;
        // printPreView(queryId: any, schemaId: any, controlType: any, qoManagerCode?: any, printIntegration?: any, queryRelativeUrl?: any, filterCondition?: any, groupType?: any, pageIndex?: any, voId?: any) {
        this.renderObject.utils.get('schema').getSchemaInfo({ schemaId: renderOption.schemaValue, queryId: renderOption.queryId, queryRelativeUrl: renderOption.queryRelativeUrl, groupType: renderOption.groupType }).subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            renderOption.schemaId = renderOption.schemaId ? renderOption.schemaId : (value && value.id ? value.id : '1');
            /** @type {?} */
            var schemaName = value && value.schemaValue && value.schemaValue.titleOption && value.schemaValue.titleOption.title
                ? value.schemaValue.titleOption.title
                : value && value.schemaName
                    ? value.schemaName
                    : '';
            renderOption.excelExportName = schemaName;
            /** @type {?} */
            var params = _this.renderObject.utils.get('parameter').createParameters(renderOption);
            if (params['schemaId'] === '1') {
                if (ServicesUtils.getServices(RtfServices.getTabId(renderOption.queryId) + 'customScheme')) {
                    params.customSchema = ServicesUtils.getServices(RtfServices.getTabId(renderOption.queryId) + 'customScheme');
                    /** @type {?} */
                    var customSchema = JSON.parse(params.customSchema);
                    if (!params.excelExportName) {
                        params.excelExportName = customSchema && customSchema.titleOption && customSchema.titleOption.title
                            ? customSchema.titleOption.title : 'export';
                    }
                }
                else {
                    _this.msgService.warning(_this.localizeService.getValue('spread.message.preview'));
                    return;
                }
            }
            if (_this.getSecurityLevel()) {
                params.excelExportName += '_' + _this.getSecurityLevel();
            }
            _this.lcpService.cloudprintService = _this.cloudprintService;
            _this.lcpService.printPreView(params, renderOption.queryRelativeUrl);
        }));
    };
    /**
     * @param {?=} queryId
     * @return {?}
     */
    RenderService.prototype.clearFirstLoadCache = /**
     * @param {?=} queryId
     * @return {?}
     */
    function (queryId) {
        /** @type {?} */
        var funcId = RtfServices.getTabId(queryId);
        /** @type {?} */
        var index = this.funcInfoCache.firstLoadCache.findIndex((/**
         * @param {?} el
         * @return {?}
         */
        function (el) { return el === funcId; }));
        if (index >= 0) {
            this.funcInfoCache.firstLoadCache.splice(index, 1);
        }
    };
    RenderService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    RenderService.ctorParameters = function () { return [
        { type: SchemaManagerService },
        { type: LcpService },
        { type: CacheService },
        { type: LoadingService },
        { type: MessagerService },
        { type: DataRenderExtendService, decorators: [{ type: Optional }] },
        { type: Injector, decorators: [{ type: Optional }] }
    ]; };
    return RenderService;
}());
export { RenderService };
if (false) {
    /** @type {?} */
    RenderService.prototype.cloudprintService;
    /** @type {?} */
    RenderService.prototype.formErrorService;
    /** @type {?} */
    RenderService.prototype.localizeService;
    /** @type {?} */
    RenderService.prototype.notifyService;
    /** @type {?} */
    RenderService.prototype.serviceCache;
    /** @type {?} */
    RenderService.prototype.columnCount;
    /** @type {?} */
    RenderService.prototype.operations;
    /** @type {?} */
    RenderService.prototype.renderObject;
    /** @type {?} */
    RenderService.prototype.spread;
    /** @type {?} */
    RenderService.prototype.funcInfoCache;
    /** @type {?} */
    RenderService.prototype.schemaManager;
    /** @type {?} */
    RenderService.prototype.lcpService;
    /**
     * @type {?}
     * @private
     */
    RenderService.prototype.cache;
    /**
     * @type {?}
     * @private
     */
    RenderService.prototype.loadService;
    /**
     * @type {?}
     * @private
     */
    RenderService.prototype.msgService;
    /**
     * @type {?}
     * @private
     */
    RenderService.prototype.dataRenderExtendService;
    /**
     * @type {?}
     * @private
     */
    RenderService.prototype.injector;
}
