// @gsp-wf/ui-comment/esm2015/lib/task-comment/task-comment.viewmodel.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { TaskCommentUIState } from './task-comment.uistate';
import { CommonComment } from '../entities/comment';
import { forkJoin } from 'rxjs';
import { HttpService } from '@ecp-caf/caf-common';
export class TaskCommentViewModel {
    /**
     * @param {?} uiState
     * @param {?} httpSvc
     */
    constructor(uiState, httpSvc) {
        this.uiState = uiState;
        this.httpSvc = httpSvc;
    }
    /**
     * @return {?}
     */
    add() {
        /** @type {?} */
        const newEntity = CommonComment.newEntity();
        newEntity.sortOrder = this.uiState.commentList.length + 1;
        this.uiState.commentList.push(newEntity);
    }
    /**
     * @return {?}
     */
    delete() {
        if (this.uiState.curComment) {
            /** @type {?} */
            const index = this.uiState.commentList.findIndex(c => c.id === this.uiState.curComment.id);
            if (index > -1) {
                this.uiState.commentList.splice(index, 1);
            }
        }
    }
    /**
     * @return {?}
     */
    save() {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/commonComments`;
        /** @type {?} */
        const deleteIds = this.uiState.originCommentIds.filter(id => this.uiState.commentList.findIndex(c => c.id === id) < 0);
        /** @type {?} */
        const addEntities = this.uiState.commentList.filter(item => this.uiState.originCommentIds.indexOf(item.id) < 0);
        /** @type {?} */
        const changesList = this.uiState.commentList.filter(item => addEntities.findIndex(aItem => aItem.id === item.id) < 0
            && this.uiState.changesCommentIds.indexOf(item.id) > -1);
        return forkJoin(deleteIds.map(id => this.httpSvc.delete(url + `/${id}`, null))
            .concat(changesList.map(change => this.httpSvc.patch(url, change)))
            .concat(addEntities.map(add => this.httpSvc.post(url, add))));
    }
}
TaskCommentViewModel.decorators = [
    { type: Injectable }
];
/** @nocollapse */
TaskCommentViewModel.ctorParameters = () => [
    { type: TaskCommentUIState },
    { type: HttpService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaskCommentViewModel.prototype.uiState;
    /**
     * @type {?}
     * @private
     */
    TaskCommentViewModel.prototype.httpSvc;
}
