// @gsp-wf/wf-task-handler/esm5/lib/components/task-comment/task-comment.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, forwardRef, Output, EventEmitter } from '@angular/core';
import { DialogComponent } from '@farris/ui-dialog';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FDropdownDirective } from '@farris/ui-dropdown';
import { UploadDialogService } from '@gsp-svc/formdoc-upload';
import { MessagerService } from '@farris/ui-messager';
import { FileUploadEntity, GspDocMetadata, GspDocUploadEntity } from '../../entity/attachment';
import { SignHereComponent } from '@gsp-wf/wf-sign';
import { AvatarComponent } from '@farris/ui-avatar';
import { DownloadService } from '@gsp-svc/file-load';
import { TaskService } from '../../service/task.service';
import { LoadingService } from '@farris/ui-loading';
import { ConfigCode } from '../../entity/task-center-config';
import { CommentManageComponent } from '../comment-manage/comment-manage.component';
var TaskCommentComponent = /** @class */ (function () {
    function TaskCommentComponent(taskService, uploadDialog, messageService, loadingService) {
        this.taskService = taskService;
        this.uploadDialog = uploadDialog;
        this.messageService = messageService;
        this.loadingService = loadingService;
        this.showAttachment = true; //是否显示附件上传，组件控制
        //是否显示附件上传，组件控制
        this.showSign = true; //是否显示签名，组件控制（不同动作可能会不显示）
        //是否显示签名，组件控制（不同动作可能会不显示）
        this.showCommonComment = true; //是否显示常用意见
        this.placeholder = '';
        this.textAreaHeight = 120; //文字框的高度
        this.attachment = new EventEmitter(); //附件
        //附件
        this.signature = new EventEmitter(); //上传签名之后，发送签名id
        //意见维护组件
        this._value = '';
        this.attachmentList = [];
        this.ifShowAttachment = false; //任务中心配置，是否显示附件上传
        //任务中心配置，是否显示附件上传
        this.ifShowSign = false; //任务中心配置，是否显示手写签名
        //临时保存上传的图片
        this.downloadService = new DownloadService();
        this.ifUseSign = false; //是否使用手写
        //是否使用手写
        this.ifUseImg = true; //是否使用上传图片
        //是否使用上传图片
        this.ifEnableSign = true; //是否默认启用签名
        //是否默认启用签名
        this.ifUploadImg = false; //是否上传过图片
        //是否上传过图片
        this.ifCanEditSign = true; //是否允许编辑签名
        //原始的签名id
        this.isShowCommonTips = false; //是否展示提示
        //是否展示提示
        this.commentList = [];
        this.btnHover = {
            //按钮hover样式变化
            commentHover: false,
            attachmentHover: false,
            signHover: false
        };
        this.onModelChange = (/**
         * @param {?} obj
         * @return {?}
         */
        function (obj) { });
        this.onModelTouched = (/**
         * @param {?} obj
         * @return {?}
         */
        function (obj) { });
    }
    Object.defineProperty(TaskCommentComponent.prototype, "TaskCenterConfig", {
        set: 
        //当前是任务列表打开还是任务办理界面打开
        /**
         * @param {?} configList
         * @return {?}
         */
        function (configList) {
            if (configList && configList.length) {
                /** @type {?} */
                var attachment = configList.find((/**
                 * @param {?} a
                 * @return {?}
                 */
                function (a) { return a.configCode == ConfigCode.If_Show_Attachment; }));
                /** @type {?} */
                var sign = configList.find((/**
                 * @param {?} a
                 * @return {?}
                 */
                function (a) { return a.configCode == ConfigCode.If_Show_Signature; }));
                /** @type {?} */
                var enableSign = configList.find((/**
                 * @param {?} a
                 * @return {?}
                 */
                function (a) { return a.configCode == ConfigCode.If_Use_Sign_By_Default; }));
                /** @type {?} */
                var editSign = configList.find((/**
                 * @param {?} a
                 * @return {?}
                 */
                function (a) { return a.configCode === ConfigCode.If_Can_Edit_Signature; }));
                attachment && attachment.currentValue === 'true' && (this.ifShowAttachment = true);
                sign && sign.currentValue === 'true' && (this.ifShowSign = true);
                enableSign && enableSign.currentValue === 'false' && (this.ifEnableSign = false);
                editSign && editSign.currentValue === 'false' && (this.ifCanEditSign = false);
                if (this.ifShowSign && this.ifEnableSign) {
                    this.getUserSign(this.curUserId);
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TaskCommentComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            if (this._value !== v) {
                this._value = v;
                this.onModelChange(v);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TaskCommentComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.commentManageComp.getConfirmSUbject().subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            _this.commentDialog && _this.commentDialog.close();
        }));
        this.commentManageComp.getCancelSubject().subscribe((/**
         * @return {?}
         */
        function () {
            _this.commentDialog && _this.commentDialog.close();
        }));
    };
    /**
     * 弹出常用意见管理框
     */
    /**
     * 弹出常用意见管理框
     * @return {?}
     */
    TaskCommentComponent.prototype.showCommentDialog = /**
     * 弹出常用意见管理框
     * @return {?}
     */
    function () {
        this.dropDownEl && this.dropDownEl.close();
        this.commentDialog.show();
    };
    /**
     * 点击常用意见，获取用户常用意见列表
     */
    /**
     * 点击常用意见，获取用户常用意见列表
     * @return {?}
     */
    TaskCommentComponent.prototype.loadComments = /**
     * 点击常用意见，获取用户常用意见列表
     * @return {?}
     */
    function () {
        var _this = this;
        this.taskService.getCommentList().subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.commentList = data;
        }));
    };
    /**
     * ControlValueAccessor必须实现的方法
     */
    /**
     * ControlValueAccessor必须实现的方法
     * @param {?} value
     * @return {?}
     */
    TaskCommentComponent.prototype.writeValue = /**
     * ControlValueAccessor必须实现的方法
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value !== this.value) {
            this.value = value;
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TaskCommentComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onModelChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TaskCommentComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onModelTouched = fn;
    };
    /**
     * 选择一条常用意见后事件
     */
    /**
     * 选择一条常用意见后事件
     * @param {?} data
     * @return {?}
     */
    TaskCommentComponent.prototype.buttonClick = /**
     * 选择一条常用意见后事件
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.dropDownEl.close();
        this.value = this.value == this.defaultComment ? data : this.value + data;
    };
    /**
     * 上传附件
     */
    /**
     * 上传附件
     * @return {?}
     */
    TaskCommentComponent.prototype.uploadAttachment = /**
     * 上传附件
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var isRepeat = false;
        this.uploadDialog.uploadFile(this.processInstanceId ? this.processInstanceId : 'gspwftaskattachment', 'wf_attachment').then((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            var e_1, _a, e_2, _b;
            try {
                for (var _c = tslib_1.__values(((/** @type {?} */ (re)))), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var i = _d.value;
                    isRepeat = false;
                    try {
                        for (var _e = tslib_1.__values(_this.attachmentList), _f = _e.next(); !_f.done; _f = _e.next()) {
                            var attach = _f.value;
                            if (attach.fileName === i.fileName) {
                                _this.messageService.info('包含重复文件' + i.fileName);
                                isRepeat = true;
                                break;
                            }
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                    if (!isRepeat) {
                        _this.attachmentList.push(i);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                }
                finally { if (e_1) throw e_1.error; }
            }
            _this.attachment.emit(_this.attachmentList);
            //执行上传附件
            /** @type {?} */
            var files = new FileUploadEntity();
            files.rootId = "wf_attachment";
            files.metadataIdList = [];
            _this.attachmentList.forEach((/**
             * @param {?} a
             * @return {?}
             */
            function (a) {
                files.metadataIdList.push(a.metadataId);
            }));
            files.metadataIdList.length && _this.taskService.uploadAttachment(files).subscribe((/**
             * @return {?}
             */
            function () { }));
        }));
    };
    /**
     * 删除附件
     */
    /**
     * 删除附件
     * @param {?} attachment
     * @return {?}
     */
    TaskCommentComponent.prototype.attachmentDelete = /**
     * 删除附件
     * @param {?} attachment
     * @return {?}
     */
    function (attachment) {
        /** @type {?} */
        var gspDocDeleteEntity = {
            rootId: "wf_attachment",
            metadataId: attachment.metadataId,
            mode: "Formal"
        };
        /** @type {?} */
        var index = this.attachmentList.indexOf(attachment);
        if (index > -1) {
            this.attachmentList.splice(index, 1);
        }
        this.attachment.emit(this.attachmentList);
        this.taskService.deleteAttachment(gspDocDeleteEntity).subscribe((/**
         * @return {?}
         */
        function () { }));
    };
    /**
    * web端打开附件预览
    */
    /**
     * web端打开附件预览
     * @param {?} metadataId
     * @return {?}
     */
    TaskCommentComponent.prototype.attachmentPreview = /**
     * web端打开附件预览
     * @param {?} metadataId
     * @return {?}
     */
    function (metadataId) {
        /** @type {?} */
        var options = {
            appType: "app",
            funcId: "",
            tabName: "文档在线预览",
            appId: "6ccb476b-f2ba-df8f-6f47-d37d100c56e2",
            appEntrance: "docviewertab",
            tabId: new Date().getTime(),
            entityParams: {
                metadataIdList: [metadataId],
                rootId: "wf_attachment",
                metadataId: null,
                viewOptions: {
                    showHeader: true,
                    showPaging: true,
                    showThumbnails: true,
                    showDownload: true,
                    showSearch: true,
                    showZoom: true,
                    showPrint: true,
                    showSecondaryToolBar: true,
                },
            },
        };
        ((/** @type {?} */ (window))).gspframeworkService.rtf.func.openMenu(options);
    };
    /**
     * 获取不同格式附件的缩略图
     */
    /**
     * 获取不同格式附件的缩略图
     * @param {?} attachmentInfo
     * @return {?}
     */
    TaskCommentComponent.prototype.imageUrl = /**
     * 获取不同格式附件的缩略图
     * @param {?} attachmentInfo
     * @return {?}
     */
    function (attachmentInfo) {
        /** @type {?} */
        var url = '/platform/runtime/wf/webapp/mobiletaskcenter/assets/accessory/';
        return url + this.taskService.createItem(attachmentInfo) + '.svg?v=20220303';
    };
    /**
     * 展示手写签名dialog
     */
    /**
     * 展示手写签名dialog
     * @return {?}
     */
    TaskCommentComponent.prototype.showSignDialog = /**
     * 展示手写签名dialog
     * @return {?}
     */
    function () {
        this.signDialog.show();
        if (!this.ifEnableSign) {
            this.getUserSign(this.curUserId);
        }
    };
    /**
     * 点击删除签名
     */
    /**
     * 点击删除签名
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToDeleteImage = /**
     * 点击删除签名
     * @return {?}
     */
    function () {
        this.signatureImage = '';
        this.signature.emit(null);
        /** @type {?} */
        var load = this.loadingService.show();
        this.taskService.addUserConfig(ConfigCode.If_Use_Sign_By_Default, 'false').subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            load.close();
        }), (/**
         * @return {?}
         */
        function () {
            load.close();
        }));
    };
    /**
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToUseSign = /**
     * @return {?}
     */
    function () {
        this.ifUseImg = false;
        this.ifUseSign = true;
        /** @type {?} */
        var canvas = this.signaturePad.canvasDiv.nativeElement.querySelector('canvas');
        canvas.height = 252;
        canvas.width = 665;
    };
    /**
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToUploadImg = /**
     * @return {?}
     */
    function () {
        this.farrisAvatar.onClick();
    };
    /**
     * 手写签名点确定
     */
    /**
     * 手写签名点确定
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToConfirm = /**
     * 手写签名点确定
     * @return {?}
     */
    function () {
        /** @type {?} */
        var image;
        if (this.signaturePad && this.signaturePad.isDraw) { //如果签过名
            image = this.signaturePad.confirm();
        }
        else if (this.ifUploadImg) { //如果上传过图片
            image = this.tmpSignatureImage;
        }
        if (image) {
            this.uploadSignatureImg(image);
        }
        else {
            this.signatureImage = this.tmpSignatureImage;
            this.signature.emit(this.originSignId);
            this.signDialog.close();
        }
        //添加启用签名配置
        /** @type {?} */
        var load = this.loadingService.show();
        this.taskService.addUserConfig(ConfigCode.If_Use_Sign_By_Default, 'true').subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            load.close();
        }), (/**
         * @return {?}
         */
        function () {
            load.close();
        }));
    };
    /**
     * 点击清空
     */
    /**
     * 点击清空
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToReset = /**
     * 点击清空
     * @return {?}
     */
    function () {
        this.ifUseSign = false;
        this.ifUseImg = true;
        this.signatureImage = '';
        this.tmpSignatureImage = '';
        this.signature.emit('');
        this.signaturePad && (this.signaturePad.isDraw = false);
        this.ifUploadImg = false;
    };
    /**
     * 点击取消
     */
    /**
     * 点击取消
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToCancel = /**
     * 点击取消
     * @return {?}
     */
    function () {
        this.signDialog.close();
    };
    /**
     * 上传图片后事件
     * event.base64 是图片信息,event.name是带后缀的图片名，event.type如image/jpeg
     */
    /**
     * 上传图片后事件
     * event.base64 是图片信息,event.name是带后缀的图片名，event.type如image/jpeg
     * @param {?} event
     * @return {?}
     */
    TaskCommentComponent.prototype.imgChange = /**
     * 上传图片后事件
     * event.base64 是图片信息,event.name是带后缀的图片名，event.type如image/jpeg
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.ifUseImg = true;
        this.ifUseSign = false;
        this.ifUploadImg = true;
        /** @type {?} */
        var image = event.base64;
        this.tmpSignatureImage = image;
        this.signatureImage = image;
    };
    /**
     * 上传签名
     * @param image
     */
    /**
     * 上传签名
     * @private
     * @param {?} image
     * @return {?}
     */
    TaskCommentComponent.prototype.uploadSignatureImg = /**
     * 上传签名
     * @private
     * @param {?} image
     * @return {?}
     */
    function (image) {
        var _this = this;
        /** @type {?} */
        var uploadSignature = new GspDocUploadEntity();
        uploadSignature.fileContent = image.split(',')[1];
        uploadSignature.filePath = this.curUserId;
        uploadSignature.mode = "Formal";
        uploadSignature.metadata = new GspDocMetadata();
        uploadSignature.metadata.id = this.taskService.createNM();
        uploadSignature.metadata.fileName = this.taskService.createNM() + ".png";
        uploadSignature.metadata.rootId = "taskcentersignature";
        uploadSignature.metadata.docType = "png";
        /** @type {?} */
        var load = this.loadingService.show();
        this.taskService.uploadSignature(uploadSignature).subscribe((/**
         * @param {?} signId
         * @return {?}
         */
        function (signId) {
            load.close();
            _this.originSignId = signId;
            _this.signature.emit(signId);
            _this.signatureImage = image;
            _this.signDialog.close();
            localStorage.setItem('gspWfTaskUserSign' + _this.curUserId + _this.taskService.getCurrentDateString(), signId);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            load.close();
            _this.messageService.error('上传签名失败');
        }));
    };
    /**
     * 获取用户的最近一次签名
     */
    /**
     * 获取用户的最近一次签名
     * @private
     * @param {?} userId
     * @return {?}
     */
    TaskCommentComponent.prototype.getUserSign = /**
     * 获取用户的最近一次签名
     * @private
     * @param {?} userId
     * @return {?}
     */
    function (userId) {
        var _this = this;
        /** @type {?} */
        var load = this.loadingService.show();
        this.taskService.getUserSignId(userId).subscribe((/**
         * @param {?} signId
         * @return {?}
         */
        function (signId) {
            load.close();
            if (signId) {
                _this.signature.emit(signId);
                _this.originSignId = signId;
                _this.signatureImage = _this.downloadService.getDownloadUrl(signId, 'taskcentersignature');
                _this.tmpSignatureImage = _this.signatureImage;
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            console.error(error);
            load.close();
        }));
    };
    //文本框值变动监听事件
    //文本框值变动监听事件
    /**
     * @param {?} e
     * @return {?}
     */
    TaskCommentComponent.prototype.valueChange = 
    //文本框值变动监听事件
    /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (this.taskService.getBytesLength(this.value) > 4000) {
            this.isShowCommonTips = true;
        }
        else {
            this.isShowCommonTips = false;
        }
        ;
    };
    TaskCommentComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-comment',
                    template: "<div\r\n  class=\"fa-approve-comment fa-approve-common-gutter\"\r\n  style=\"margin-bottom: 14px\"\r\n>\r\n  <!-- \u8F93\u5165\u6846 -->\r\n  <div class=\"fa-approve-comment--content\">\r\n    <textarea\r\n      class=\"form-control content-textarea\"\r\n      style=\"font-size: 13px\"\r\n      [(ngModel)]=\"value\"\r\n      (ngModelChange)=\"valueChange($event)\"\r\n      [attr.placeholder]=\"placeholder\"\r\n      [style.height.px]=\"textAreaHeight\"\r\n    ></textarea>\r\n  </div>\r\n  <div\r\n    class=\"fa-approve-comment--footer wf-approve-comment--footer\"\r\n    *ngIf=\"isShowCommonTips\"\r\n  >\r\n    <div class=\"wf-feature-warning\"></div>\r\n    <span\r\n      ><span class=\"wf-feature-span\">{{\r\n        \"comment.static.comment\" | translate\r\n      }}</span\r\n      >{{ \"comment.static.tooLongTips\" | translate }}</span\r\n    >\r\n  </div>\r\n  <!-- \u8F93\u5165\u6846\u4E0B\u65B9\u4E00\u6392\u6309\u94AE -->\r\n  <div class=\"fa-approve-comment--footer\">\r\n    <div\r\n      *ngIf=\"showAttachment || showSign || showCommonComment\"\r\n      class=\"approve-comment--text-btns\"\r\n    >\r\n      <!-- \u5E38\u7528\u610F\u89C1 -->\r\n      <div\r\n        *ngIf=\"showCommonComment\"\r\n        class=\"text-btns-item\"\r\n        (click)=\"loadComments()\"\r\n        (mouseover)=\"btnHover.commentHover = true\"\r\n        (mouseleave)=\"btnHover.commentHover = false\"\r\n        [ngClass]=\"btnHover.commentHover && 'text-primary'\"\r\n      >\r\n        <div fDropdown #dropdownEl=\"fDropdown\">\r\n          <div fDropdownToggle>\r\n            <span class=\"f-icon f-icon-message\"></span>\r\n            <span>\r\n              {{ \"comment.static.inputLabel\" | translate }}\r\n            </span>\r\n          </div>\r\n          <div class=\"dropdown-menu task-drop-down-menu\" fDropdownMenu>\r\n            <li\r\n              class=\"dropdown-item\"\r\n              *ngFor=\"let comment of commentList\"\r\n              (click)=\"buttonClick(comment.message)\"\r\n            >\r\n              <span>{{ comment.message }}</span>\r\n            </li>\r\n            <div class=\"dropdown-divider\"></div>\r\n            <!-- \u610F\u89C1\u7BA1\u7406 -->\r\n            <div\r\n              class=\"dropdown-item\"\r\n              style=\"display: flex; align-items: center\"\r\n              (click)=\"showCommentDialog()\"\r\n            >\r\n              <span class=\"f-icon f-icon-home-setup\"></span>\r\n              <span style=\"padding-left: 6px\">{{\r\n                \"comment.static.manage\" | translate\r\n              }}</span>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <!-- \u4E0A\u4F20\u9644\u4EF6 -->\r\n      <div\r\n        class=\"text-btns-item\"\r\n        *ngIf=\"ifShowAttachment && showAttachment\"\r\n        (click)=\"uploadAttachment()\"\r\n        (mouseover)=\"btnHover.attachmentHover = true\"\r\n        (mouseleave)=\"btnHover.attachmentHover = false\"\r\n        [ngClass]=\"btnHover.attachmentHover && 'text-primary'\"\r\n        title=\"{{ 'comment.attachment.tips' | translate }}\"\r\n      >\r\n        <span class=\"f-icon f-icon-attachment\"></span>\r\n        <span class=\"text-btns-item--text\">\r\n          {{ \"comment.static.attachmentUpload\" | translate }}\r\n        </span>\r\n      </div>\r\n      <!-- \u624B\u5199\u7B7E\u540D -->\r\n      <div\r\n        class=\"text-btns-item\"\r\n        *ngIf=\"showSign && ifShowSign && ifCanEditSign && curUserId\"\r\n        title=\"{{ 'comment.signature.tips' | translate }}\"\r\n        (click)=\"showSignDialog()\"\r\n        (mouseover)=\"btnHover.signHover = true\"\r\n        (mouseleave)=\"btnHover.signHover = false\"\r\n        [ngClass]=\"btnHover.signHover && 'text-primary'\"\r\n      >\r\n        <span class=\"f-icon f-icon-edit\"></span>\r\n        <span>{{ \"comment.signature.title\" | translate }}</span>\r\n      </div>\r\n      <!-- \u624B\u5199\u7B7E\u540D\u9884\u89C8 -->\r\n      <div\r\n        *ngIf=\"showSign && ifShowSign && signatureImage\"\r\n        class=\"signatureImage\"\r\n      >\r\n        <img [src]=\"signatureImage\" style=\"width: 70px; height: 26px\" />\r\n        <div\r\n          *ngIf=\"ifCanEditSign\"\r\n          class=\"attachment-delete\"\r\n          (click)=\"clickToDeleteImage()\"\r\n          title=\"{{ 'comment.signature.disable' | translate }}\"\r\n        >\r\n          <span class=\"f-icon f-icon-x\"></span>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<!-- \u4E0A\u4F20\u540E\u7684\u9644\u4EF6\u9884\u89C8 -->\r\n<div *ngIf=\"attachmentList?.length > 0\" style=\"margin-bottom: 20px\">\r\n  <div\r\n    class=\"task-comment-attachment\"\r\n    *ngFor=\"let attachment of attachmentList\"\r\n  >\r\n    <!-- \u9644\u4EF6\u540D\u79F0 -->\r\n    <div class=\"task-comment-attachment-title\">\r\n      <img class=\"task-comment-attachment-img\" [src]=\"imageUrl(attachment)\" />\r\n      <span>{{ attachment.fileName }}</span>\r\n    </div>\r\n    <div style=\"display: flex; align-items: center\">\r\n      <!-- \u9644\u4EF6\u5927\u5C0F -->\r\n      <div *ngIf=\"attachment.fileSize\" class=\"task-comment-attachment-size\">\r\n        {{ attachment.fileSize }}\r\n      </div>\r\n      <!-- \u67E5\u770B -->\r\n      <div\r\n        *ngIf=\"taskOrigin == 'TaskForm'\"\r\n        class=\"task-comment-attach-icon\"\r\n        (click)=\"attachmentPreview(attachment.metadataId)\"\r\n        title=\"{{ 'comment.signature.view' | translate }}\"\r\n      >\r\n        <span class=\"f-icon f-icon-enclosure_browse\"></span>\r\n      </div>\r\n      <!-- \u5220\u9664 -->\r\n      <div\r\n        class=\"task-comment-attach-icon\"\r\n        (click)=\"attachmentDelete(attachment)\"\r\n        title=\"{{ 'comment.signature.delete' | translate }}\"\r\n      >\r\n        <span class=\"f-icon f-icon-enclosure_delete\"></span>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<!-- \u5E38\u7528\u610F\u89C1\u7EF4\u62A4 -->\r\n<farris-dialog\r\n  #commentDialog\r\n  [width]=\"740\"\r\n  [height]=\"440\"\r\n  [buttons]=\"commentManageComp.buttons\"\r\n  [title]=\"'comment.static.title' | translate\"\r\n>\r\n  <lib-comment-manage #commentManageComp></lib-comment-manage>\r\n</farris-dialog>\r\n\r\n<!-- \u7B7E\u540D\u5F39\u6846 -->\r\n<farris-dialog\r\n  #signDialog\r\n  [buttons]=\"signButtons\"\r\n  [width]=\"710\"\r\n  [height]=\"420\"\r\n  [title]=\"'comment.signature.dialogTitle' | translate\"\r\n  [resizable]=\"false\"\r\n  [showMaxButton]=\"false\"\r\n>\r\n  <wf-sign-here\r\n    #wfSignHere\r\n    [title]=\"''\"\r\n    [ngStyle]=\"!ifUseSign ? { display: 'none' } : null\"\r\n  ></wf-sign-here>\r\n  <div *ngIf=\"ifUseImg && tmpSignatureImage\" class=\"signature-img-default\">\r\n    <img [src]=\"tmpSignatureImage\" style=\"height: 252px\" />\r\n  </div>\r\n  <div *ngIf=\"ifUseImg && !tmpSignatureImage\" class=\"signature-img-div\">\r\n    <div class=\"signature-img-main\">\r\n      <div style=\"margin-bottom: 28px\">\r\n        <img\r\n          src=\"/platform/runtime/common/web/@gsp-wf/wf-task-handler/assets/img/signIcon.svg?v=20210623\"\r\n        />\r\n      </div>\r\n      <div class=\"signature-img-text\">\r\n        <span>{{ \"comment.signature.note1\" | translate }}</span>\r\n        <span class=\"signature-text-action\" (click)=\"clickToUseSign()\">{{\r\n          \"comment.signature.sign\" | translate\r\n        }}</span>\r\n        <span>{{ \"comment.signature.note2\" | translate }}</span>\r\n        <span class=\"signature-text-action\" (click)=\"clickToUploadImg()\">{{\r\n          \"comment.signature.upload\" | translate\r\n        }}</span>\r\n        <span>{{ \"comment.signature.note3\" | translate }}</span>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n\r\n<!-- \u624B\u5199\u7B7E\u540D\u6309\u94AE -->\r\n<ng-template #signButtons>\r\n  <button class=\"btn btn-secondary\" (click)=\"clickToCancel()\">\r\n    {{ \"comment.static.cancel\" | translate }}\r\n  </button>\r\n  <button class=\"btn btn-secondary mr-2\" (click)=\"clickToReset()\">\r\n    {{ \"comment.signature.reset\" | translate }}\r\n  </button>\r\n  <button class=\"btn btn-primary mr-2\" (click)=\"clickToConfirm()\">\r\n    {{ \"comment.signature.ok\" | translate }}\r\n  </button>\r\n</ng-template>\r\n\r\n<!-- \u4F7F\u7528\u5934\u50CF\u7EC4\u4EF6\u505A\u56FE\u7247\u4E0A\u4F20 -->\r\n<farris-avatar\r\n  #farrisAvatar\r\n  (imgChange)=\"imgChange($event)\"\r\n  style=\"display: none\"\r\n></farris-avatar>\r\n",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaskCommentComponent; })),
                            multi: true
                        },],
                    styles: [":host ::ng-deep .dropdown .btn{padding:0}:host ::ng-deep .dropdown .btn-secondary{box-shadow:none}:host ::ng-deep .dropdown .btn-secondary:focus{box-shadow:none}:host ::ng-deep .dropdown .btn-secondary:hover{color:#388fff}.signatureImage:hover .attachment-delete{display:block}.fa-approve-common-gutter{margin:0 0 12px}.fa-approve-comment .approve-comment--title{display:inline-block;color:#667580;border-radius:4px 4px 0 0;height:26px;padding:0 9px;position:relative;background:#f7f7f7;cursor:pointer}.fa-approve-comment .approve-comment--title:after{content:\"\";height:100%;position:absolute;right:-10px;top:0;bottom:0;border-right:5px solid transparent;border-left:5px solid #f7f7f7;border-bottom:13px solid #f7f7f7;border-top:13px solid transparent}.fa-approve-comment .approve-comment--title .fa-btn-like-dropdown{font-size:12px;padding:2px 0}.fa-approve-comment .fa-approve-comment--content .content-textarea{border-color:#dcdcdc;border-radius:0 0 2px 2px}.fa-approve-comment .fa-approve-comment--footer{font-size:13px;color:#667580;background:#f7f7f7;align-items:center;padding:5px 0}.fa-approve-comment .approve-comment--text-btns{display:flex;flex-direction:row;padding:0 5px}.fa-approve-comment .approve-comment--text-btns .text-btns-item{display:flex;flex-direction:row;align-items:center;font-size:12px;line-height:18px;color:#667580;margin-right:12px;cursor:pointer}.fa-approve-comment .approve-comment--text-btns .text-btns-item .f-icon{margin-right:4px}li{list-style:none}.attachment-delete{float:right;cursor:pointer;display:none;text-align:right;padding-right:10px}.attachment-delete:hover{color:#4297fa}.signature-img-div{display:flex;justify-content:center;align-items:center;height:100%;background:#fafcfd;border:1px solid #d8dfed;border-radius:8px;margin:0 10px}.signature-img-main{display:flex;flex-direction:column;justify-content:center;align-items:center}.signature-img-text{font-size:13px;color:rgba(106,111,121,.85)}.signature-text-action{color:#529efe;padding:0 2px;cursor:pointer}.signature-img-default{display:flex;justify-content:center;align-items:center;height:inherit}.wf-feature-warning{display:inline-block;width:16px;height:16px;line-height:1px;vertical-align:text-bottom;background:url(/platform/runtime/wf/webapp/mobiletaskcenter/assets/img/icon_validator.png)}.wf-approve-comment--footer{font-weight:700;align-items:center;font-family:\"PingFang SC\",\"Lantinghei SC\",\"Helvetica Neue\",Arial,\"Microsoft YaHei\",\"WenQuanYi Micro Hei\",\"Heiti SC\",\"Segoe UI\",sans-serif}.wf-feature-span{color:#f60;padding:0 2px}.task-comment-attachment{display:flex;justify-content:space-between;align-items:center;background:rgba(249,252,255,.87);border:0 solid rgba(229,232,237,.71);border-radius:3px;margin-top:10px;padding:4px;cursor:pointer;font-size:13px}.task-comment-attachment-title{display:flex;align-items:center;font-size:12px}.task-comment-attachment-img{width:20px;margin-right:10px}.task-comment-attachment-size{font-size:12px;padding:0 10px;color:rgba(45,47,51,.45)}.task-comment-attach-icon{margin-left:10px;height:16px;width:16px;color:var(--primary);border-radius:50%;display:flex;align-items:center;justify-content:center;font-size:10px;background:#e8f2ff;cursor:pointer}.task-drop-down-menu{max-height:200px;overflow:auto}"]
                }] }
    ];
    /** @nocollapse */
    TaskCommentComponent.ctorParameters = function () { return [
        { type: TaskService },
        { type: UploadDialogService },
        { type: MessagerService },
        { type: LoadingService }
    ]; };
    TaskCommentComponent.propDecorators = {
        showAttachment: [{ type: Input }],
        showSign: [{ type: Input }],
        showCommonComment: [{ type: Input }],
        comment: [{ type: Input }],
        placeholder: [{ type: Input }],
        processInstanceId: [{ type: Input }],
        textAreaHeight: [{ type: Input }],
        defaultComment: [{ type: Input }],
        curUserId: [{ type: Input }],
        taskOrigin: [{ type: Input }],
        TaskCenterConfig: [{ type: Input }],
        attachment: [{ type: Output }],
        signature: [{ type: Output }],
        commentDialog: [{ type: ViewChild, args: ['commentDialog',] }],
        signDialog: [{ type: ViewChild, args: ['signDialog',] }],
        dropDownEl: [{ type: ViewChild, args: ['dropdownEl',] }],
        signaturePad: [{ type: ViewChild, args: ['wfSignHere',] }],
        farrisAvatar: [{ type: ViewChild, args: ['farrisAvatar',] }],
        commentManageComp: [{ type: ViewChild, args: ['commentManageComp',] }]
    };
    return TaskCommentComponent;
}());
export { TaskCommentComponent };
if (false) {
    /** @type {?} */
    TaskCommentComponent.prototype.showAttachment;
    /** @type {?} */
    TaskCommentComponent.prototype.showSign;
    /** @type {?} */
    TaskCommentComponent.prototype.showCommonComment;
    /** @type {?} */
    TaskCommentComponent.prototype.comment;
    /** @type {?} */
    TaskCommentComponent.prototype.placeholder;
    /** @type {?} */
    TaskCommentComponent.prototype.processInstanceId;
    /** @type {?} */
    TaskCommentComponent.prototype.textAreaHeight;
    /** @type {?} */
    TaskCommentComponent.prototype.defaultComment;
    /** @type {?} */
    TaskCommentComponent.prototype.curUserId;
    /** @type {?} */
    TaskCommentComponent.prototype.taskOrigin;
    /** @type {?} */
    TaskCommentComponent.prototype.attachment;
    /** @type {?} */
    TaskCommentComponent.prototype.signature;
    /** @type {?} */
    TaskCommentComponent.prototype.commentDialog;
    /** @type {?} */
    TaskCommentComponent.prototype.signDialog;
    /** @type {?} */
    TaskCommentComponent.prototype.dropDownEl;
    /** @type {?} */
    TaskCommentComponent.prototype.signaturePad;
    /** @type {?} */
    TaskCommentComponent.prototype.farrisAvatar;
    /** @type {?} */
    TaskCommentComponent.prototype.commentManageComp;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype._value;
    /** @type {?} */
    TaskCommentComponent.prototype.attachmentList;
    /** @type {?} */
    TaskCommentComponent.prototype.ifShowAttachment;
    /** @type {?} */
    TaskCommentComponent.prototype.ifShowSign;
    /** @type {?} */
    TaskCommentComponent.prototype.signatureImage;
    /** @type {?} */
    TaskCommentComponent.prototype.tmpSignatureImage;
    /** @type {?} */
    TaskCommentComponent.prototype.downloadService;
    /** @type {?} */
    TaskCommentComponent.prototype.ifUseSign;
    /** @type {?} */
    TaskCommentComponent.prototype.ifUseImg;
    /** @type {?} */
    TaskCommentComponent.prototype.ifEnableSign;
    /** @type {?} */
    TaskCommentComponent.prototype.ifUploadImg;
    /** @type {?} */
    TaskCommentComponent.prototype.ifCanEditSign;
    /** @type {?} */
    TaskCommentComponent.prototype.originSignId;
    /** @type {?} */
    TaskCommentComponent.prototype.isShowCommonTips;
    /** @type {?} */
    TaskCommentComponent.prototype.commentList;
    /** @type {?} */
    TaskCommentComponent.prototype.btnHover;
    /** @type {?} */
    TaskCommentComponent.prototype.onModelChange;
    /** @type {?} */
    TaskCommentComponent.prototype.onModelTouched;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype.taskService;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype.uploadDialog;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype.loadingService;
}
