// @qdp/formular/esm5/lib/service/entitydata.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/entitydata.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
var EntityDataService = /** @class */ (function () {
    function EntityDataService() {
        this.isFinish = false;
        this.formularInfo = {};
    }
    /**
     * @param {?} stepTye
     * @return {?}
     */
    EntityDataService.prototype.setStepType = /**
     * @param {?} stepTye
     * @return {?}
     */
    function (stepTye) {
        this.stepType = stepTye;
    };
    /**
     * @return {?}
     */
    EntityDataService.prototype.getStepType = /**
     * @return {?}
     */
    function () {
        return this.stepType;
    };
    /**
     * @param {?} isFinish
     * @param {?} formularInfo
     * @return {?}
     */
    EntityDataService.prototype.setFormularState = /**
     * @param {?} isFinish
     * @param {?} formularInfo
     * @return {?}
     */
    function (isFinish, formularInfo) {
        this.isFinish = isFinish;
        this.formularInfo = formularInfo;
    };
    /**
     * @return {?}
     */
    EntityDataService.prototype.getFormularState = /**
     * @return {?}
     */
    function () {
        return { isFinish: this.isFinish, formularInfo: this.formularInfo };
    };
    EntityDataService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    EntityDataService.ctorParameters = function () { return []; };
    return EntityDataService;
}());
export { EntityDataService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    EntityDataService.prototype.isFinish;
    /**
     * @type {?}
     * @private
     */
    EntityDataService.prototype.formularInfo;
    /**
     * @type {?}
     * @private
     */
    EntityDataService.prototype.stepType;
}
