// @qdp/formular/esm5/lib/service/restful.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/service/restful.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
var RestfulService = /** @class */ (function () {
    /**
     * 构造函数
     * @param httpClient http客户端
     */
    function RestfulService(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * 发送GET请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    /**
     * 发送GET请求
     * @param {?} uri 请求地址
     * @param {?=} params 请求参数
     * @param {?=} options
     * @return {?}
     */
    RestfulService.prototype.get = /**
     * 发送GET请求
     * @param {?} uri 请求地址
     * @param {?=} params 请求参数
     * @param {?=} options
     * @return {?}
     */
    function (uri, params, options) {
        return this.request(uri, 'GET', params, options);
    };
    /**
     * 发送POST请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    /**
     * 发送POST请求
     * @param {?} uri 请求地址
     * @param {?} body
     * @param {?=} params 请求参数
     * @param {?=} options
     * @return {?}
     */
    RestfulService.prototype.put = /**
     * 发送POST请求
     * @param {?} uri 请求地址
     * @param {?} body
     * @param {?=} params 请求参数
     * @param {?=} options
     * @return {?}
     */
    function (uri, body, params, options) {
        /** @type {?} */
        var optionsWithBody = this.addBody(options, body);
        return this.request(uri, 'PUT', params, optionsWithBody);
    };
    /**
     * 发送PUT请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    /**
     * 发送PUT请求
     * @param {?} uri 请求地址
     * @param {?} body
     * @param {?=} params 请求参数
     * @param {?=} options
     * @return {?}
     */
    RestfulService.prototype.post = /**
     * 发送PUT请求
     * @param {?} uri 请求地址
     * @param {?} body
     * @param {?=} params 请求参数
     * @param {?=} options
     * @return {?}
     */
    function (uri, body, params, options) {
        /** @type {?} */
        var optionsWithBody = this.addBody(options, body);
        return this.request(uri, 'POST', params, optionsWithBody);
    };
    /**
     * 发送DELETE请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    /**
     * 发送DELETE请求
     * @param {?} uri 请求地址
     * @param {?=} params 请求参数
     * @param {?=} options
     * @return {?}
     */
    RestfulService.prototype.delete = /**
     * 发送DELETE请求
     * @param {?} uri 请求地址
     * @param {?=} params 请求参数
     * @param {?=} options
     * @return {?}
     */
    function (uri, params, options) {
        return this.request(uri, 'DELETE', params, options);
    };
    /**
     * 发送请求
     * @param uri 请求地址
     * @param params 请求参数
     */
    /**
     * 发送请求
     * @private
     * @param {?} uri 请求地址
     * @param {?} method
     * @param {?} params 请求参数
     * @param {?=} options
     * @return {?}
     */
    RestfulService.prototype.request = /**
     * 发送请求
     * @private
     * @param {?} uri 请求地址
     * @param {?} method
     * @param {?} params 请求参数
     * @param {?=} options
     * @return {?}
     */
    function (uri, method, params, options) {
        if (options === void 0) { options = {}; }
        options = options || {};
        if (params) {
            /** @type {?} */
            var httpParams = this.buildParams(params);
            options.params = httpParams;
        }
        /** @type {?} */
        var methodName = (/** @type {?} */ (method));
        /** @type {?} */
        var result = this.httpClient.request(methodName, uri, options);
        // if (result && result.returnValue) {result = result.returnValue};
        return result;
    };
    /**
     * 构造Http请求参数
     * @param params 请求参数
     */
    /**
     * 构造Http请求参数
     * @private
     * @param {?} params 请求参数
     * @return {?}
     */
    RestfulService.prototype.buildParams = /**
     * 构造Http请求参数
     * @private
     * @param {?} params 请求参数
     * @return {?}
     */
    function (params) {
        /** @type {?} */
        var parameters = {};
        for (var key in params) {
            if (params.hasOwnProperty(key) && params[key]) {
                /** @type {?} */
                var value = params[key].toString();
                parameters[key] = value;
            }
        }
        return parameters;
    };
    /**
     * @private
     * @param {?} options
     * @param {?} body
     * @return {?}
     */
    RestfulService.prototype.addBody = /**
     * @private
     * @param {?} options
     * @param {?} body
     * @return {?}
     */
    function (options, body) {
        options = options || {};
        /** @type {?} */
        var mergedOptions = Object.assign(options, { body: body });
        return mergedOptions;
    };
    RestfulService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    RestfulService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return RestfulService;
}());
export { RestfulService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    RestfulService.prototype.httpClient;
}
