// @ecp-caf/common-structure/esm2015/lib/convertor/utils/data-type-convert-util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { StructuredReferType } from '../../define/data-type/structured-type-refer-type/structured-refer-type';
import { StructuredType } from '../../define/data-type/structured-type/structured-type';
export class DataTypeConvertUtil {
    /**
     * @param {?} datatype
     * @param {?} ctxt
     * @return {?}
     */
    static convertToObject(datatype, ctxt) {
        /** @type {?} */
        let result = new Object();
        if (datatype instanceof StructuredType) {
            result['kind'] = "RefStructuredType";
            result["isRefer"] = true;
            result["refId"] = datatype.id;
        }
        else {
            /** @type {?} */
            var kind = datatype.getKind();
            result["kind"] = kind;
            /** @type {?} */
            var serializer = ctxt.getDataTypeSerializerFactory().get(kind);
            result = serializer.serialize(datatype, ctxt);
        }
        return result;
    }
    /**
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    static initFromObject(obj, ctxt) {
        /** @type {?} */
        let result;
        if (obj['isRefer'] != null) {
            result = new StructuredReferType();
            ((/** @type {?} */ (result))).refId = obj['refId'];
        }
        else {
            /** @type {?} */
            var kind = obj["kind"];
            /** @type {?} */
            var serializer = ctxt.getDataTypeSerializerFactory().get(kind);
            result = serializer.deserialize(obj, ctxt);
        }
        return result;
        // switch (obj["kind"]) {
        //     case DataTypeConst.DEFAULT_PRIMITIVE:
        //         let primitiveKindConvertor = new PrimitiveTypeConvertor();
        //         return primitiveKindConvertor.initFromObject(obj);
        //     case DataTypeConst.DEFAULT_COLLECTION:
        //         let collectionConvertor = new CollectionTypeConvertor();
        //         return collectionConvertor.initFromObject(obj);
        //     case DataTypeConst.DEFAULT_MAP:
        //         let mapConvertor = new MapTypeConvertor();
        //         return mapConvertor.initFromObject(obj);
        //     case DataTypeConst.DEFAULT_STRUCTURED:
        //         let structureConvertor = new StructuredTypeConvertor();
        //         return structureConvertor.initFromObject(obj);
        //     case DataTypeConst.DEFAULT_ENTITY:
        //         let entityConvertor = new StructuredTypeConvertor();
        //         return entityConvertor.initFromObject(obj);
        //     case DataTypeConst.DEFAULT_VALUE_OBJECT:
        //         let voConvertor = new StructuredTypeConvertor();
        //         return voConvertor.initFromObject(obj);
        //     case DataTypeConst.DEFAULT_DATA_TRANSFER_OBJECT:
        //         let dtoConvertor = new StructuredTypeConvertor();
        //         return dtoConvertor.initFromObject(obj);
        //     case DataTypeConst.DEFAULT_ENUMERATION:
        //         let enumerationConvertor = new StructuredTypeConvertor();
        //         return enumerationConvertor.initFromObject(obj);
        //     case DataTypeConst.STRUCTURED_REF:
        //         let structureRefConvertor = new StructuredTypeReferConvertor();
        //         return structureRefConvertor.initFromObject(obj);
        //     default:
        //         throw new Error(`数据类型不支持的${obj["kind"]}类型！`);
        // }
    }
}
