// @farris/ui-sidebar/esm2015/lib/sidebar.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class FarrisSidebarService {
    constructor() {
        this.isOpenSubject = new Subject();
    }
    /**
     * @param {?} message
     * @return {?}
     */
    sendIsOpen(message) {
        this.isOpenSubject.next(message);
    }
    /**
     * @return {?}
     */
    getIsOpen() {
        return this.isOpenSubject.asObservable();
    }
}
FarrisSidebarService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
FarrisSidebarService.ctorParameters = () => [];
/** @nocollapse */ FarrisSidebarService.ngInjectableDef = i0.defineInjectable({ factory: function FarrisSidebarService_Factory() { return new FarrisSidebarService(); }, token: FarrisSidebarService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    FarrisSidebarService.prototype.isOpenSubject;
}
