// @farris/ui-datagrid/esm5/lib/components/loading.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
/** @type {?} */
var loading_style_ring = "<svg width=\"100%\" height=\"100%\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 100 100\" preserveAspectRatio=\"xMidYMid\">\n  <circle cx=\"50\" cy=\"50\" fill=\"none\" r=\"30\" stroke=\"#2c7dff\" stroke-width=\"10\"></circle>\n  <circle cx=\"50\" cy=\"50\" fill=\"none\" r=\"30\" stroke=\"#d0eaff\" stroke-width=\"10\" stroke-linecap=\"square\" transform=\"rotate(178.821 50 50)\">\n      <animateTransform attributeName=\"transform\" type=\"rotate\"\n          calcMode=\"linear\" values=\"0 50 50;180 50 50;720 50 50\" keyTimes=\"0;0.5;1\" dur=\"1.6s\" begin=\"0s\"\n          repeatCount=\"indefinite\"></animateTransform>\n      <animate attributeName=\"stroke-dasharray\" calcMode=\"linear\"\n          values=\"9.42477796076938 179.0707812546182;150.79644737231007 37.6991118430775;9.42477796076938 179.0707812546182\"\n          keyTimes=\"0;0.5;1\" dur=\"1.6\" begin=\"0s\" repeatCount=\"indefinite\"></animate>\n  </circle>\n</svg>";
var DataGridLoadingComponent = /** @class */ (function () {
    function DataGridLoadingComponent(domSanitizer) {
        this.domSanitizer = domSanitizer;
    }
    /**
     * @return {?}
     */
    DataGridLoadingComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.loading) {
            this.loading = this.domSanitizer.bypassSecurityTrustHtml(this.loading);
        }
        else {
            this.loading = this.domSanitizer.bypassSecurityTrustHtml(loading_style_ring);
        }
    };
    DataGridLoadingComponent.decorators = [
        { type: Component, args: [{
                    selector: 'datagrid-loading',
                    template: "\n    <div style=\"width:100%;height:100%;position:absolute;top:0;left: 0;background:rgba(255,255,255,0);z-index: 90000000;pointer-events: unset;\">\n        <div style=\"width: 50px;height: 50px;position: relative;top: 50%;margin-top: -25px;left: 50%;margin-left: -25px;\"\n        [innerHTML]=\"loading\" >\n        </div>\n    </div>\n    "
                }] }
    ];
    /** @nocollapse */
    DataGridLoadingComponent.ctorParameters = function () { return [
        { type: DomSanitizer }
    ]; };
    DataGridLoadingComponent.propDecorators = {
        loading: [{ type: Input }]
    };
    return DataGridLoadingComponent;
}());
export { DataGridLoadingComponent };
if (false) {
    /** @type {?} */
    DataGridLoadingComponent.prototype.loading;
    /**
     * @type {?}
     * @private
     */
    DataGridLoadingComponent.prototype.domSanitizer;
}
