// @gsp-wf/wf-process-editor/esm5/lib/domain/entities/biz-component.entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ProcessDeUtil } from '../process-de-util';
var BizComponentEntity = /** @class */ (function () {
    function BizComponentEntity(name, metadataid, method) {
        this.id = ProcessDeUtil.GenerateElementId();
        this.name = name;
        this.metadataId = metadataid;
        this.methodCode = method;
        this.isForCompensation = false;
    }
    return BizComponentEntity;
}());
export { BizComponentEntity };
if (false) {
    /** @type {?} */
    BizComponentEntity.prototype.id;
    /** @type {?} */
    BizComponentEntity.prototype.name;
    /** @type {?} */
    BizComponentEntity.prototype.flowFormKey;
    /** @type {?} */
    BizComponentEntity.prototype.bizActivityId;
    /** @type {?} */
    BizComponentEntity.prototype.componentLocation;
    /** @type {?} */
    BizComponentEntity.prototype.executionTime;
    /** @type {?} */
    BizComponentEntity.prototype.metadataId;
    /** @type {?} */
    BizComponentEntity.prototype.methodCode;
    /** @type {?} */
    BizComponentEntity.prototype.methodName;
    /** @type {?} */
    BizComponentEntity.prototype.actualParameters;
    /** @type {?} */
    BizComponentEntity.prototype.owner;
    /** @type {?} */
    BizComponentEntity.prototype.ownerType;
    /** @type {?} */
    BizComponentEntity.prototype.isForCompensation;
    /** @type {?} */
    BizComponentEntity.prototype.condition;
    /** @type {?} */
    BizComponentEntity.prototype.scopeType;
}
/** @enum {number} */
var ComponentLocation = {
    Process: 0,
    Activity: 1,
};
export { ComponentLocation };
ComponentLocation[ComponentLocation.Process] = 'Process';
ComponentLocation[ComponentLocation.Activity] = 'Activity';
/** @enum {string} */
var OwnerType = {
    Process: 'Process',
    Activity: 'Activity',
    Schema: 'Schema',
};
export { OwnerType };
