// @gsp-wf/wf-process-editor/esm5/lib/domain/entities/organization-participant-expression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var OrganizationParticipantExpression = /** @class */ (function () {
    function OrganizationParticipantExpression() {
    }
    return OrganizationParticipantExpression;
}());
export { OrganizationParticipantExpression };
if (false) {
    /** @type {?} */
    OrganizationParticipantExpression.prototype.id;
    /** @type {?} */
    OrganizationParticipantExpression.prototype.name;
    /** @type {?} */
    OrganizationParticipantExpression.prototype.actionScope;
    /** @type {?} */
    OrganizationParticipantExpression.prototype.organizationLevel;
    /** @type {?} */
    OrganizationParticipantExpression.prototype.organizationId;
    /** @type {?} */
    OrganizationParticipantExpression.prototype.organizationCode;
    /** @type {?} */
    OrganizationParticipantExpression.prototype.organizationName;
    /** @type {?} */
    OrganizationParticipantExpression.prototype.includeChildren;
    /** @type {?} */
    OrganizationParticipantExpression.prototype.parentsUntil;
    /** @type {?} */
    OrganizationParticipantExpression.prototype.tag;
    /** @type {?} */
    OrganizationParticipantExpression.prototype.condition;
}
/** @enum {string} */
var ActionScope = {
    /**
     * 指派
     */
    AssignUser: 'AssignUser',
    /**
     * 移交
     */
    Transfer: 'Transfer',
    /**
     * 征询
     */
    AddSignFront: 'AddSignFront',
    /**
     * 加签
     */
    AddSignBehind: 'AddSignBehind',
    /**
     * 知会
     */
    Notify: "Notify",
};
export { ActionScope };
/** @enum {string} */
var OrganizationLevel = {
    /**
     * 部门
     */
    DEPARTMENT: 'DEPARTMENT',
    /**
     * 单位
     */
    UNIT: 'UNIT',
    /**
    * 默认值
    */
    DEFAULT: 'DEFAULT',
    VARORGANIZATION: 'VARORGANIZATION',
};
export { OrganizationLevel };
