// @farris/kendo-binding/esm2015/lib/directives/end-edit/input_end_edit.directive.js
import { Directive, ElementRef } from '@angular/core';
import { AppContext } from '@farris/devkit';
import { AbstractEndEdit } from './abstract_end_edit';
class InputEndEditDirective extends AbstractEndEdit {
    constructor(appContext, elementRef) {
        super(appContext);
        this.appContext = appContext;
        this.elementRef = elementRef;
    }
    ngOnInit() {
        if (this.elementRef) {
            this.nativeElement = this.elementRef.nativeElement;
        }
    }
    /**
     * 结束编辑
     * @param message 消息
     */
    endEdit(message) {
        if (this.nativeElement) {
            const parent = this.nativeElement.parentNode || this.nativeElement.parentElement;
            const inputs = parent && Array.from(parent.getElementsByTagName('input')) || [];
            if (inputs && inputs.length > 0) {
                inputs.forEach(input => {
                    if (input && typeof (input.blur) === 'function') {
                        input.blur();
                    }
                });
            }
        }
    }
}
InputEndEditDirective.decorators = [
    { type: Directive, args: [{
                selector: '[input-end-edit]'
            },] }
];
/** @nocollapse */
InputEndEditDirective.ctorParameters = () => [
    { type: AppContext },
    { type: ElementRef }
];
export { InputEndEditDirective };
