// @farris/appointment-calendar/esm2015/lib/views/base.view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Injector, Input, NgZone, Output, ViewChild } from '@angular/core';
import { CommonUtils, OverLayHiddenService } from '@farris/ui-common';
import { AppointmentCalendarComponent } from '../appointment-calendar.component';
export class BaseViewComponent {
    /**
     * @param {?} injector
     * @param {?} ngzone
     * @param {?} calendarRef
     */
    constructor(injector, ngzone, calendarRef) {
        this.injector = injector;
        this.ngzone = ngzone;
        this.calendarRef = calendarRef;
        this.selectId = '';
        this.itemClick = new EventEmitter();
        this.itemDblClick = new EventEmitter();
        this.scrollChange = new EventEmitter();
        this.activeItem = null;
        this.currentMousePointX = 0;
        this.currentMousePointY = 0;
        this.isClick = true;
        this.overlaySer = new OverLayHiddenService();
        this.utils = this.injector.get(CommonUtils);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.setActiveItem();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.ngzone.runOutsideAngular((/**
         * @return {?}
         */
        () => {
            this.bodyRef.nativeElement.addEventListener('scroll', (/**
             * @param {?} e
             * @return {?}
             */
            (e) => {
                /** @type {?} */
                const x = this.bodyRef.nativeElement.scrollLeft;
                /** @type {?} */
                const y = this.bodyRef.nativeElement.scrollTop;
                this.headerRef.nativeElement.scrollLeft = this.bodyRef.nativeElement.scrollLeft;
                this.calendarRef.closeDetailPanel();
                this.scrollChange.emit({ x, y });
            }));
        }));
        if (this.overlaySer) {
            this.overlaySer.registerMouseEvent(this.bodyRef.nativeElement, (/**
             * @param {?} e
             * @return {?}
             */
            (e) => {
                if (this.calendarRef.detailRef) {
                    if (this.calendarRef.detailRef.el.nativeElement.contains(e.target)) {
                        return;
                    }
                    if (e.type !== 'mousewheel' && this.isClick) {
                        if (this.calendarRef.viewType === 'week') {
                            if (e.target.closest('li') && e.target.closest('li').classList.contains('week-day-subject')) {
                                return;
                            }
                        }
                        if (this.calendarRef.viewType === 'day') {
                            if (e.target.closest('div.item-content')) {
                                return;
                            }
                        }
                    }
                    if (e.target.classList.contains('farris-modal') || e.target.closest('.farris-modal')) {
                        if (e.target.closest('farris-messager') && e.target.nodeName !== 'BUTTON') {
                            return;
                        }
                    }
                    this.calendarRef.closeDetailPanel();
                }
            }), { mouseUp: true });
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.overlaySer) {
            this.overlaySer.destory(this.bodyRef.nativeElement);
            this.overlaySer = null;
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.selectId && !changes.selectId.isFirstChange()) {
            this.setActiveItem();
        }
        if (changes.data && !changes.data.isFirstChange()) {
            this.setActiveItem();
        }
    }
    /**
     * @param {?=} id
     * @return {?}
     */
    setActiveItem(id) { }
    /**
     * @param {?} data
     * @return {?}
     */
    getRoomData(data) {
        /** @type {?} */
        const roomid = this.utils.getValue(this.calendarRef.reserveInfoFields.placeid, data);
        /** @type {?} */
        const room = this.data.find((/**
         * @param {?} n
         * @return {?}
         */
        (n) => n.id === roomid));
        if (room) {
            return room.data;
        }
        return null;
    }
    /**
     * @param {?} $event
     * @param {?} item
     * @return {?}
     */
    onSubjectClick($event, item) {
        $event.stopPropagation();
        if (!this.isClick) {
            this.calendarRef.closeDetailPanel();
            return;
        }
        if (this.calendarRef.viewType === 'week') {
            $event.target.scrollIntoViewIfNeeded();
        }
        if (this['clickTimer']) {
            clearTimeout(this['clickTimer']);
        }
        this['clickTimer'] = setTimeout((/**
         * @return {?}
         */
        () => {
            this.activeItem = item;
            /** @type {?} */
            const room = this.getRoomData(item.data);
            this.itemClick.emit({ item, $event, room });
        }), this.calendarRef._CLICK_DELAY_);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    ondblClick($event) {
        if (this['clickTimer']) {
            clearTimeout(this['clickTimer']);
        }
        this.activeItem = $event.item;
        /** @type {?} */
        const room = this.getRoomData($event.item.data);
        $event['room'] = room;
        this.itemDblClick.emit($event);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onMouseDown($event) {
        this.currentMousePointX = Math.floor($event.screenX);
        this.currentMousePointY = Math.floor($event.screenY);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onMouseUp($event) {
        this.isClick = this.currentMousePointX == Math.floor($event.screenX) && this.currentMousePointY == Math.floor($event.screenY);
    }
}
BaseViewComponent.decorators = [
    { type: Component, args: [{
                selector: 'appointment-base-view',
                template: ''
            }] }
];
/** @nocollapse */
BaseViewComponent.ctorParameters = () => [
    { type: Injector },
    { type: NgZone },
    { type: AppointmentCalendarComponent }
];
BaseViewComponent.propDecorators = {
    data: [{ type: Input }],
    selectId: [{ type: Input }],
    itemClick: [{ type: Output }],
    itemDblClick: [{ type: Output }],
    headerRef: [{ type: ViewChild, args: ['header',] }],
    bodyRef: [{ type: ViewChild, args: ['body',] }],
    scrollChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    BaseViewComponent.prototype.data;
    /** @type {?} */
    BaseViewComponent.prototype.selectId;
    /** @type {?} */
    BaseViewComponent.prototype.itemClick;
    /** @type {?} */
    BaseViewComponent.prototype.itemDblClick;
    /** @type {?} */
    BaseViewComponent.prototype.headerRef;
    /** @type {?} */
    BaseViewComponent.prototype.bodyRef;
    /** @type {?} */
    BaseViewComponent.prototype.scrollChange;
    /** @type {?} */
    BaseViewComponent.prototype.activeItem;
    /**
     * @type {?}
     * @private
     */
    BaseViewComponent.prototype.overlaySer;
    /**
     * @type {?}
     * @protected
     */
    BaseViewComponent.prototype.utils;
    /**
     * @type {?}
     * @private
     */
    BaseViewComponent.prototype.currentMousePointX;
    /**
     * @type {?}
     * @private
     */
    BaseViewComponent.prototype.currentMousePointY;
    /**
     * @type {?}
     * @private
     */
    BaseViewComponent.prototype.isClick;
    /**
     * @type {?}
     * @protected
     */
    BaseViewComponent.prototype.injector;
    /**
     * @type {?}
     * @protected
     */
    BaseViewComponent.prototype.ngzone;
    /** @type {?} */
    BaseViewComponent.prototype.calendarRef;
}
