// @farris/appointment-calendar/esm2015/lib/views/day-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Injector, Input, NgZone, Output, Renderer2, ViewEncapsulation, } from "@angular/core";
import { AppointmentCalendarComponent } from "../appointment-calendar.component";
import { flatten } from "lodash-es";
import { BaseViewComponent } from "./base.view.component";
export class DayViewComponent extends BaseViewComponent {
    /**
     * @param {?} injector
     * @param {?} ngzone
     * @param {?} calendarRef
     */
    constructor(injector, ngzone, calendarRef) {
        super(injector, ngzone, calendarRef);
        this.reservationTitle = '预订';
        this.cellClick = new EventEmitter();
        this.cells = new Array(23);
        this.render = null;
        this.render = this.injector.get(Renderer2, null);
    }
    /**
     * @param {?} item
     * @param {?=} typ
     * @return {?}
     */
    getCustomCls(item, typ = "cls") {
        /** @type {?} */
        let result = null;
        if (this.customCls) {
            const { cls, styles } = this.customCls(item);
            if (!cls && !styles && typ === "cls") {
                result = this.customCls(item);
            }
            else {
                if (typ == "cls") {
                    result = cls || null;
                }
                else {
                    result = styles || null;
                }
            }
        }
        // if (result && typ === 'styles') {
        //     Object.keys(result).forEach((n: any) => {
        //         if (result[n].indexOf('important') === -1) {
        //             result[n] = result[n] +' !important' ;
        //         }
        //     });
        // }
        return result;
    }
    /**
     * @param {?=} id
     * @return {?}
     */
    setActiveItem(id) {
        /** @type {?} */
        let _id = id;
        if (!id) {
            _id = this.selectId;
        }
        if (_id) {
            this.activeItem = flatten(this.data.map((/**
             * @param {?} n
             * @return {?}
             */
            (n) => n.reserves))).find((/**
             * @param {?} n
             * @return {?}
             */
            (n) => n.id == _id));
        }
    }
    /**
     * @param {?} $event
     * @param {?} cellIndex
     * @param {?} roomInfo
     * @return {?}
     */
    onCellClick($event, cellIndex, roomInfo) {
        /** @type {?} */
        const _h = ('' + cellIndex).padStart(2, '0');
        /** @type {?} */
        const _endH = ('' + (cellIndex + 1)).padStart(2, '0');
        /** @type {?} */
        const p = {
            start: `${this.calendarRef["currentDateStr"]} ${_h}:00:00`,
            end: `${this.calendarRef["currentDateStr"]} ${_endH}:00:00`,
            place: roomInfo.data,
        };
        this.cellClick.emit(p);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onMouseEnter($event) {
        $event.stopPropagation();
        /** @type {?} */
        const parentCell = $event.target['parentElement'].querySelector('.addnew');
        if (parentCell) {
            parentCell.style.opacity = 0;
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onMouseLeave($event) {
        $event.stopPropagation();
        /** @type {?} */
        const parentCell = $event.target['parentElement'].querySelector('.addnew');
        if (parentCell) {
            parentCell.removeAttribute('style');
        }
    }
}
DayViewComponent.decorators = [
    { type: Component, args: [{
                selector: "day-view",
                template: "<div class=\"header\" #header>\r\n\r\n    <div class=\"header-row\" [style.width.px]=\"calendarRef.rowWidth\">\r\n\r\n        <div class=\"header-cell roomname fixed-left\" [style.width.px]=\"calendarRef.placeNameWidth\">\r\n            <span class=\"time ml-2\">{{calendarRef.placeTitle}}</span>\r\n        </div>\r\n\r\n        <div class=\"header-cell time-cell\" style=\"width: 50px\">\r\n        </div>\r\n        <div class=\"header-cell time-cell\" >\r\n            <span class=\"time\">00:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">01:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">02:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">03:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">04:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">05:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">06:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">07:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">08:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">09:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">10:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">11:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">12:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">13:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">14:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">15:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">16:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">17:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">18:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">19:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">20:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">21:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\">\r\n            <span class=\"time\">22:00</span>\r\n        </div>\r\n        <div class=\"header-cell time-cell\"  style=\"width: 100px\">\r\n            <span class=\"time\">23:00</span>\r\n        </div>\r\n           \r\n    </div>\r\n\r\n</div>\r\n<div class=\"body win11Scroll nobtn\" #body>\r\n\r\n    <div class=\"room-row\" *ngFor=\"let room of data\" [style.height.px]=\"calendarRef.rowHeight\"  [style.width.px]=\"calendarRef.rowWidth\"  [attr.id]=\"'row_'+room.id\">\r\n\r\n        <div class=\"room-cell roomname fixed-left\" [style.width.px]=\"calendarRef.placeNameWidth\">\r\n            <span *ngIf=\"!calendarRef.placeNameTemplate\">{{ room.data[calendarRef.placeNameField] }}</span>\r\n            <ng-container [ngTemplateOutlet]=\"calendarRef.placeNameTemplate\" [ngTemplateOutletContext]=\"{ $implicit: {data: room.data}, item: room.data }\"></ng-container>\r\n        </div>\r\n        <div class=\"room-cell time-cell\" style=\"width: 50px\">\r\n        </div>\r\n        <div class=\"room-cell time-cell\">\r\n            <div class=\"item-content\" *ngFor=\"let item of room?.reserves\" \r\n                [item-dblclick]=\"item\" (dblclickHandler)=\"ondblClick($event)\"\r\n                [style.width.px]=\"item.position?.width\" [style.left.px]=\"item.position?.left\"\r\n                (click)=\"onSubjectClick($event, item)\" \r\n                (mousedown)=\"onMouseDown($event)\" (mouseup)=\"onMouseUp($event)\"\r\n                 (mouseenter)=\"onMouseEnter($event)\" (mouseleave)=\"onMouseLeave($event)\">\r\n                <div class=\"item-content-wrap\" [ngClass]=\"customCls && getCustomCls(item)\" [ngStyle]=\"customCls && getCustomCls(item, 'styles')\" [class.selected]=\"activeItem && activeItem.id === item.id\">\r\n                    <div class=\"room-subject p-2 gray\" *ngIf=\"!calendarRef.contentTemplate\">\r\n                        <dl> \r\n                            <dt>{{ calendarRef.utils.getValue(calendarRef.reserveInfoFields.title, item.data) }}</dt>\r\n                            <dd class=\"other-info\">\r\n                                <span>\r\n                                    {{ calendarRef.utils.getValue(calendarRef.reserveInfoFields.startDate, item.data) |dateformat: 'HH:mm'}}\r\n                                    ~\r\n                                    {{ calendarRef.utils.getValue(calendarRef.reserveInfoFields.endDate, item.data) |dateformat: 'HH:mm'}}\r\n                                </span>\r\n                            </dd>\r\n                        </dl>\r\n                    </div>\r\n                    <ng-container [ngTemplateOutlet]=\"calendarRef.contentTemplate\" [ngTemplateOutletContext]=\"{ $implicit: {data: item.data}, item: item.data }\"></ng-container>\r\n                </div>\r\n            </div>\r\n\r\n\r\n            <ng-container *ngIf=\"!(room.disableCells && room.disableCells.indexOf(0) > -1)\">\r\n\r\n                <div class=\"addnew\" (click)=\"onCellClick($event, 0, room)\">\r\n                    <span class=\"f-icon f-icon-add\" style=\"position: relative;\r\n                    top: 1px;margin-right: 3px;\"></span> {{reservationTitle}}\r\n                </div>\r\n\r\n            </ng-container>\r\n        </div>\r\n        <div class=\"room-cell time-cell\" *ngFor=\"let c of cells; let i= index\" [class.isUsed]=\"room.disableCells && room.disableCells.indexOf(i+1) > -1 \" >\r\n\r\n\r\n            <ng-container *ngIf=\"!(room.disableCells && room.disableCells.indexOf(i+1) > -1)\">\r\n\r\n                <div class=\"addnew\" (click)=\"onCellClick($event, (i + 1), room)\">\r\n                    <span class=\"f-icon f-icon-add\" style=\"position: relative;\r\n                    top: 1px;\r\n                    margin-right: 3px;\"></span> {{reservationTitle}}\r\n                </div>\r\n\r\n            </ng-container>\r\n\r\n        </div>\r\n    </div>\r\n\r\n</div>\r\n",
                encapsulation: ViewEncapsulation.None,
                styles: [`
        .rtv-container .main .room-row .time-cell .addnew{
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  border: 1px solid #2a87ff;
  background: #edf5ff;
  cursor: pointer;
  opacity: 0;
  transition: all 0.12s ease;
  color: #045dd3;
  font-size: 14px;
  border-radius: 4px;
}

.rtv-container .main .room-row .time-cell:hover>.addnew {
  opacity: 1;
}

        `]
            }] }
];
/** @nocollapse */
DayViewComponent.ctorParameters = () => [
    { type: Injector },
    { type: NgZone },
    { type: AppointmentCalendarComponent }
];
DayViewComponent.propDecorators = {
    customCls: [{ type: Input }],
    reservationTitle: [{ type: Input }],
    cellClick: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DayViewComponent.prototype.customCls;
    /** @type {?} */
    DayViewComponent.prototype.reservationTitle;
    /** @type {?} */
    DayViewComponent.prototype.cellClick;
    /** @type {?} */
    DayViewComponent.prototype.cells;
    /** @type {?} */
    DayViewComponent.prototype.render;
}
