// @farris/appointment-calendar/esm2015/lib/views/toolbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Injector, Input, Output, ViewEncapsulation } from '@angular/core';
import { DateTimeHelperService } from '@farris/ui-common/date';
import { addDays, subDays } from 'date-fns';
import { AppointmentCalendarComponent } from '../appointment-calendar.component';
export class ToolbarComponent {
    /**
     * @param {?} dateHlp
     * @param {?} injector
     * @param {?} calendarRef
     */
    constructor(dateHlp, injector, calendarRef) {
        this.dateHlp = dateHlp;
        this.injector = injector;
        this.calendarRef = calendarRef;
        this.dateChanged = new EventEmitter();
        this.placeChanged = new EventEmitter();
        this.currentDate = '';
        this.currentDateRange = '';
        this.currentWeekDays = [];
        this.currentViewType = 'day';
        this.places = [];
        this.placeIdField = 'id';
        this.placeNameField = 'name';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.initDate();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.rooms && !changes.rooms.isFirstChange()) {
            this.places = [{ [this.placeIdField]: '', [this.placeNameField]: '全部场所' }, ...this.places];
        }
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() { }
    /**
     * @private
     * @return {?}
     */
    initDate() {
        if (this.currentViewType === 'week' && !this.currentDateRange) {
            this.currentWeekDays = this.dateHlp.getWeekDate({ baselineDate: new Date() });
            this.currentDateRange = this.currentWeekDays[0].dateStr + '~' + this.currentWeekDays[6].dateStr;
        }
        else {
            if (!this.currentDate) {
                this.currentDate = this.dateHlp.formatTo(new Date(), 'yyyy-MM-dd');
            }
        }
        this._dateChanged();
    }
    /**
     * @param {?} $event
     * @param {?} viewType
     * @return {?}
     */
    changeCalendarView($event, viewType) {
        if ($event) {
            $event.stopPropagation();
        }
        this.currentViewType = viewType;
        this.initDate();
    }
    /**
     * @param {?} $event
     * @param {?} typ
     * @return {?}
     */
    onDateButtonClicked($event, typ) {
        /** @type {?} */
        let baselineDate = new Date();
        if (this.currentViewType === 'day') {
            if (this.currentDate) {
                baselineDate = new Date(this.currentDate);
            }
            /** @type {?} */
            let newDateStr = this.dateHlp.formatTo(new Date(), 'yyyy-MM-dd');
            if (typ === -1) {
                /** @type {?} */
                const d = this.dateHlp.formatTo(subDays(baselineDate, 1), 'yyyy-MM-dd');
                newDateStr = this.dateHlp.formatTo(d);
            }
            else if (typ === 1) {
                /** @type {?} */
                const d = this.dateHlp.formatTo(addDays(baselineDate, 1), 'yyyy-MM-dd');
                newDateStr = this.dateHlp.formatTo(d);
            }
            this.currentDate = newDateStr;
        }
        else {
            if (this.currentWeekDays && this.currentWeekDays.length) {
                baselineDate = this.currentWeekDays[4].date;
            }
            // 下周
            if (typ === 1) {
                this.currentWeekDays = this.dateHlp.getWeekDate({ baselineDate, range: 7 });
            }
            else if (typ === -1) { // 上周
                this.currentWeekDays = this.dateHlp.getWeekDate({ baselineDate, range: -7 });
            }
            else { // 当前周
                baselineDate = new Date();
                this.currentWeekDays = this.dateHlp.getWeekDate({ baselineDate });
            }
            this.currentDateRange = this.currentWeekDays[0].dateStr + '~' + this.currentWeekDays[6].dateStr;
            // console.log(this.currentWeekDays);
        }
        this._dateChanged();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onDateChanged($event) {
        /** @type {?} */
        let isChanged = false;
        if (this.currentViewType == 'week') {
            /** @type {?} */
            const _date = $event.split('~')[0];
            this.currentWeekDays = this.dateHlp.getWeekDate({ baselineDate: new Date(_date) });
            isChanged = this.currentDateRange !== $event;
        }
        else {
            isChanged = this.currentDate !== $event;
        }
        if (isChanged) {
            this._dateChanged($event);
        }
    }
    /**
     * @private
     * @param {?=} newDate
     * @return {?}
     */
    _dateChanged(newDate) {
        if (newDate) {
            if (this.currentViewType === 'day') {
                this.currentDate = newDate;
            }
            else {
                this.currentDateRange = newDate;
            }
        }
        this.dateChanged.emit({
            viewType: this.currentViewType,
            dateValue: this.currentViewType === 'day' ? this.currentDate : this.currentDateRange,
            weekDays: this.currentWeekDays
        });
        this.calendarRef.closeDetailPanel();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onPlaceSelectChanged($event) {
        const { data } = $event;
        this.placeChanged.emit({
            viewType: this.currentViewType,
            dateValue: this.currentViewType === 'day' ? this.currentDate : this.currentDateRange,
            weekDays: this.currentWeekDays,
            room: data.id ? data : null
        });
    }
}
ToolbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'toolbar',
                template: "<div class=\"toolbar\" style=\"overflow-x: auto ;\">\r\n\r\n    <div class=\"view-type \">\r\n        <button class=\"btn-day\" type=\"button\" [class.btn-active]=\"currentViewType === 'day'\" (click)=\"changeCalendarView($event, 'day')\">{{'appointment.day'|locale}}</button>\r\n        <button class=\"btn-day\" type=\"button\" [class.btn-active]=\"currentViewType === 'week'\" (click)=\"changeCalendarView($event, 'week')\">{{'appointment.week'|locale}}</button>\r\n    </div>\r\n    <span class=\"line\"></span>\r\n    <div class=\"btns\">\r\n        <button class=\"btn-today\" (click)=\"onDateButtonClicked($event, 0)\">{{ currentViewType === 'day'? ('appointment.today'|locale ): ('appointment.thisWeek'|locale)}}</button>\r\n\r\n        <div class=\"btn-group\">\r\n            <button class=\"btn-prev\" (click)=\"onDateButtonClicked($event, -1)\" title=\"{{ currentViewType === 'day'? ('appointment.prevDay'|locale): ('appointment.prevWeek'|locale) }}\">\r\n                <i class=\"f-icon f-icon-arrow-chevron-left\"></i>\r\n            </button>\r\n            <button class=\"btn-next\" (click)=\"onDateButtonClicked($event, 1)\" title=\"{{ currentViewType === 'day'? ('appointment.nextDay'|locale): ('appointment.nextWeek'|locale) }}\">\r\n                <i class=\"f-icon f-icon-arrow-chevron-right\"></i>\r\n            </button>\r\n        </div>\r\n    </div>\r\n\r\n    <div class=\"inputs d-flex\">\r\n        <div class=\"datebox mr-4\" [style.width.px]=\"currentViewType === 'day'? 180 : 260\">\r\n            <farris-datepicker [ngModel]=\"currentDate\" [editable]=\"false\"\r\n                dateFormat=\"{{'appointment.dateFormat'|locale}}\" returnFormat=\"yyyy-MM-dd\" \r\n                placeholder=\"\u8BF7\u9009\u62E9\" *ngIf=\"currentViewType === 'day'\" [showClear]=\"false\"\r\n                (valueChange)=\"onDateChanged($event.returnFormatted)\"></farris-datepicker>\r\n            <farris-datepicker [ngModel]=\"currentDateRange\"  [editable]=\"false\" [dateRange]=\"true\"\r\n                dateFormat=\"{{'appointment.dateFormat'|locale}}\" returnFormat=\"yyyy-MM-dd\" [showType]=\"4\" \r\n                *ngIf=\"currentViewType === 'week'\" [showClear]=\"false\"\r\n                (valueChange)=\"onDateChanged($event.returnFormatted)\"></farris-datepicker>\r\n        </div>\r\n        <div class=\"roomlist-input\" style=\"width: 176px; display: none;\">\r\n            <farris-combo-list #cmb1 \r\n                [placeholder]=\"'appointment.datePlace'|locale\"\r\n                [enableClear]=\"true\"\r\n                [idField]=\"placeIdField\"\r\n                [textField]=\"placeNameField\"\r\n                [data]=\"places\" \r\n                [panelHeight]=\"260\"\r\n                [readonly]=\"false\"\r\n                [disabled]=\"false\"\r\n                [editable]=\"true\"\r\n                (selectChange)=\"onPlaceSelectChanged($event)\">\r\n            </farris-combo-list>\r\n        </div>\r\n    </div>\r\n\r\n</div>",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
ToolbarComponent.ctorParameters = () => [
    { type: DateTimeHelperService },
    { type: Injector },
    { type: AppointmentCalendarComponent }
];
ToolbarComponent.propDecorators = {
    dateChanged: [{ type: Output }],
    placeChanged: [{ type: Output }],
    currentViewType: [{ type: Input }],
    places: [{ type: Input }],
    placeIdField: [{ type: Input }],
    placeNameField: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    ToolbarComponent.prototype.dateChanged;
    /** @type {?} */
    ToolbarComponent.prototype.placeChanged;
    /** @type {?} */
    ToolbarComponent.prototype.currentDate;
    /** @type {?} */
    ToolbarComponent.prototype.currentDateRange;
    /** @type {?} */
    ToolbarComponent.prototype.currentWeekDays;
    /** @type {?} */
    ToolbarComponent.prototype.currentViewType;
    /** @type {?} */
    ToolbarComponent.prototype.places;
    /** @type {?} */
    ToolbarComponent.prototype.placeIdField;
    /** @type {?} */
    ToolbarComponent.prototype.placeNameField;
    /**
     * @type {?}
     * @private
     */
    ToolbarComponent.prototype.dateHlp;
    /**
     * @type {?}
     * @private
     */
    ToolbarComponent.prototype.injector;
    /** @type {?} */
    ToolbarComponent.prototype.calendarRef;
}
