// @farris/appointment-calendar/esm2015/lib/views/week-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Injector, Input, NgZone, ViewEncapsulation } from '@angular/core';
import { AppointmentCalendarComponent } from '../appointment-calendar.component';
import { flatten } from 'lodash-es';
import { BaseViewComponent } from './base.view.component';
export class WeekViewComponent extends BaseViewComponent {
    /**
     * @param {?} injector
     * @param {?} ngzone
     * @param {?} calendarRef
     */
    constructor(injector, ngzone, calendarRef) {
        super(injector, ngzone, calendarRef);
        this.weekDays = [];
    }
    /**
     * @param {?=} id
     * @return {?}
     */
    setActiveItem(id) {
        /** @type {?} */
        let _id = id;
        if (!id) {
            _id = this.selectId;
        }
        if (_id) {
            /** @type {?} */
            const arr = flatten(this.data.map((/**
             * @param {?} n
             * @return {?}
             */
            (n) => n.reserves.map((/**
             * @param {?} t
             * @return {?}
             */
            t => t.items)))));
            this.activeItem = flatten(arr).find((/**
             * @param {?} n
             * @return {?}
             */
            n => n.id == _id));
        }
    }
}
WeekViewComponent.decorators = [
    { type: Component, args: [{
                selector: 'week-view',
                template: "<div class=\"header\" #header>\r\n\r\n    <div class=\"header-row\" [attr.style]=\"'width: ' +  width + 'px!important' | safe: 'style'\">\r\n\r\n        <div class=\"header-cell roomname fixed-left\" [style.width.px]=\"calendarRef.placeNameWidth\">\r\n            <span class=\"time ml-2\">{{calendarRef.placeTitle}}</span>\r\n        </div>\r\n\r\n        <div style=\"flex: 7;display: flex;\">\r\n            <div class=\"header-cell time-cell\" *ngFor=\"let item of weekDays\" title=\"{{ item.dateStr }}  {{ 'appointment.weeks.' + item.week | locale}}\" [style.width.px]=\"calendarRef.weekCellWidth\">\r\n                <span class=\"rtv-week-day\" [class.active]=\"item.isToday\">{{item.day}}</span> <span>\r\n                    {{ 'appointment.weeks.' + item.week | locale}}</span>\r\n            </div>\r\n        </div>\r\n       \r\n           \r\n    </div>\r\n\r\n</div>\r\n<div class=\"body win11Scroll nobtn\" #body style=\"padding-bottom: 15px\">\r\n\r\n    <div class=\"room-row\" *ngFor=\"let room of data\" [attr.style]=\"'width: ' +  width + 'px!important; height:' + calendarRef.weekRowHeight + 'px' | safe: 'style'\">\r\n\r\n        <div class=\"room-cell roomname fixed-left\" [style.width.px]=\"calendarRef.placeNameWidth\">\r\n            <span *ngIf=\"!calendarRef.placeNameTemplate\">{{ room.data[calendarRef.placeNameField] }}</span>\r\n            <ng-container [ngTemplateOutlet]=\"calendarRef.placeNameTemplate\" [ngTemplateOutletContext]=\"{ $implicit: {data: room.data}, item: room.data }\"></ng-container>        \r\n        </div>\r\n        <div style=\"flex: 7;display: flex;\">\r\n            <div class=\"room-cell time-cell win11Scroll nobtn\" *ngFor=\"let res of room?.reserves\" style=\"align-items: start;\" [style.width.px]=\"calendarRef.weekCellWidth\">\r\n                <div style=\"width: 100%;overflow: hidden;\" class=\"p-1\" *ngIf=\"res.items && res.items.length\">\r\n                    <ul style=\"list-style: none;word-break: keep-all;\">\r\n                        <li class=\"week-day-subject mb-1\" *ngFor=\"let item of res.items\" \r\n                            [class.selected]=\"activeItem && activeItem.id === item.id\" (mousedown)=\"onMouseDown($event)\" (mouseup)=\"onMouseUp($event)\"\r\n                            (click)=\"onSubjectClick($event, item)\" [item-dblclick]=\"item\" (dblclickHandler)=\"ondblClick($event)\">\r\n                            <div style=\"display: flex;flex-direction: row;\" class=\"rtv-week-day-item\">\r\n                                <div style=\"flex-grow: 1;overflow: hidden;text-overflow: ellipsis;white-space: nowrap;\" title=\"{{ calendarRef.reserveInfoFields.title | getvalue: item.data }}\">\r\n                                    {{ calendarRef.utils.getValue(calendarRef.reserveInfoFields.title || 'title', item.data) }}\r\n                                </div>\r\n                                <div  style=\"flex: none;text-align: right; padding-left: 3px;\">\r\n                                    {{ item.data[calendarRef.reserveInfoFields.startDate] | dateformat: 'HH:mm'}}\r\n                                </div>\r\n                            </div>                        \r\n                        </li>\r\n                    </ul>\r\n                </div>\r\n            </div>\r\n        </div>\r\n\r\n    </div>\r\n\r\n</div>\r\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
WeekViewComponent.ctorParameters = () => [
    { type: Injector },
    { type: NgZone },
    { type: AppointmentCalendarComponent }
];
WeekViewComponent.propDecorators = {
    weekDays: [{ type: Input }],
    width: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    WeekViewComponent.prototype.weekDays;
    /** @type {?} */
    WeekViewComponent.prototype.width;
}
