// @qdp/common/esm5/lib/dnd-list/directives/dnd-nodrag.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostBinding, HostListener } from '@angular/core';
var DndNodragDirective = /** @class */ (function () {
    function DndNodragDirective() {
    }
    /**
     * @return {?}
     */
    DndNodragDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    Object.defineProperty(DndNodragDirective.prototype, "draggable", {
        get: /**
         * @return {?}
         */
        function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    DndNodragDirective.prototype.onDragStart = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event = event.originalEvent || event;
        if (!event._dndHandle) {
            // If a child element already reacted to dragstart and set a dataTransfer object, we will
            // allow that. For example, this is the case for user selections inside of input elements.
            if (!(event.dataTransfer.types && event.dataTransfer.types.length)) {
                event.preventDefault();
            }
            event.stopPropagation();
        }
    };
    /**
     * Stop propagation of dragend events, otherwise dnd-moved might be triggered and the element
     * would be removed.
     */
    /**
     * Stop propagation of dragend events, otherwise dnd-moved might be triggered and the element
     * would be removed.
     * @param {?} event
     * @return {?}
     */
    DndNodragDirective.prototype.onDragEnd = /**
     * Stop propagation of dragend events, otherwise dnd-moved might be triggered and the element
     * would be removed.
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event = event.originalEvent || event;
        if (!event._dndHandle) {
            event.stopPropagation();
        }
    };
    DndNodragDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[dndNodrag]',
                },] }
    ];
    /** @nocollapse */
    DndNodragDirective.ctorParameters = function () { return []; };
    DndNodragDirective.propDecorators = {
        draggable: [{ type: HostBinding, args: ['draggable',] }],
        onDragStart: [{ type: HostListener, args: ['dragstart', ['$event'],] }],
        onDragEnd: [{ type: HostListener, args: ['dragend', ['$event'],] }]
    };
    return DndNodragDirective;
}());
export { DndNodragDirective };
