// @farris/ui-property-panel/esm5/lib/components/property-panel/property-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, ElementRef, Output, EventEmitter, ViewChildren } from '@angular/core';
import { PropertyItemListComponent } from '../property-item-list/property-item-list.component';
var PropertyPanelComponent = /** @class */ (function () {
    function PropertyPanelComponent() {
        this.isPersitOpenState = false; // 是否持有面板的隐藏显示状态
        // 是否持有面板的隐藏显示状态
        this.isShowPanel = false; // isPersitOpenState=true时，控制面板是否隐藏显示
        // isPersitOpenState=true时，控制面板是否隐藏显示
        this.closePropertyPanel = new EventEmitter(); // isPersitOpenState=true时，抛出面板的关闭事件
        // 属性类型
        this.propertyData = {}; // 属性值
        // 属性值
        this.propertyChanged = new EventEmitter(); // 属性变更事件
        // 属性变更事件
        this.submitModal = new EventEmitter(); // 模态框属性变更事件
        // 模态框属性变更事件
        this.showCloseBtn = true;
        // 面板标题，为适配国际化
        this.panelTitle = '属性面板';
        this.isOpen = true;
    }
    /**
     * @return {?}
     */
    PropertyPanelComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.width) {
            this.eastDiv.nativeElement.style.width = this.width + 'px';
        }
        else {
            this.eastDiv.nativeElement.style.width = 'inherit';
        }
        if (this.height) {
            this.eastDiv.nativeElement.style.height = this.height + 'px';
        }
        else {
            this.eastDiv.nativeElement.style.height = '100%';
        }
        if (this.isPersitOpenState) {
            this.isOpen = this.isShowPanel;
        }
    };
    /**
     *  更改面板的隐藏显示状态，只有在isPersitOpenState=true时生效
     * @param simpleChanges
     */
    /**
     *  更改面板的隐藏显示状态，只有在isPersitOpenState=true时生效
     * @param {?} simpleChanges
     * @return {?}
     */
    PropertyPanelComponent.prototype.ngOnChanges = /**
     *  更改面板的隐藏显示状态，只有在isPersitOpenState=true时生效
     * @param {?} simpleChanges
     * @return {?}
     */
    function (simpleChanges) {
        if (simpleChanges.isShowPanel && this.isPersitOpenState) {
            this.isOpen = this.isShowPanel;
        }
    };
    /**
     * @return {?}
     */
    PropertyPanelComponent.prototype.refreshPanel = /**
     * @return {?}
     */
    function () {
        this.itemLists.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.refresh(); }));
    };
    /**
     * @param {?} item
     * @return {?}
     */
    PropertyPanelComponent.prototype.changeStatus = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        if (this.propertyConfig.length < 2) {
            return;
        }
        if (!item.status || item.status === 'open') {
            item.status = 'closed';
        }
        else {
            item.status = 'open';
        }
    };
    // 隐藏面板
    // 隐藏面板
    /**
     * @return {?}
     */
    PropertyPanelComponent.prototype.collapse = 
    // 隐藏面板
    /**
     * @return {?}
     */
    function () {
        // isPersitOpenState=true时,由外部确定状态
        if (!this.isPersitOpenState) {
            this.isOpen = false;
        }
        this.closePropertyPanel.emit();
    };
    /**
     * 抛出属性变更事件
     * @param changeObject
     */
    /**
     * 抛出属性变更事件
     * @param {?} changeObject
     * @return {?}
     */
    PropertyPanelComponent.prototype._valueChanged = /**
     * 抛出属性变更事件
     * @param {?} changeObject
     * @return {?}
     */
    function (changeObject) {
        this.propertyChanged.emit(changeObject);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    PropertyPanelComponent.prototype._submitModal = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.submitModal.emit($event);
    };
    PropertyPanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-property-panel',
                    template: "<div #east [style.display]=\"isOpen?'block':'none'\">\r\n    <div class=\"propertyPanel panel d-flex flex-column\">\r\n\r\n        <div class=\"title d-block\">\r\n            <div class=\"title-label\">\r\n                <span>{{panelTitle|| '\u5C5E\u6027\u9762\u677F'}}</span>\r\n            </div>\r\n            <div class=\"title-actions\">\r\n                <div class=\"monaco-toolbar\">\r\n                    <div class=\"monaco-action-bar animated\">\r\n                        <ul class=\"actions-container\" role=\"toolbar\">\r\n                            <li class=\"action-item\" (click)=\"collapse()\" *ngIf=\"showCloseBtn\">\r\n                                <a class=\"hide-panel\"></a>\r\n                            </li>\r\n                        </ul>\r\n                    </div>\r\n                </div>\r\n            </div>\r\n        </div>\r\n\r\n        <div #propertyPanel class=\"panel-body\" [style.height]=\"(height - 35) + 'px'\">\r\n            <perfect-scrollbar [config]=\"{suppressScrollX: true}\">\r\n                <ul class=\"property-grid\">\r\n                    <li *ngFor=\"let category of propertyConfig\">\r\n                        <ng-container *ngIf=\"!category.hide\">\r\n                            <span *ngIf=\"!category.hideTitle\"\r\n                                [class]=\"category.status === 'closed' ? 'group-label collapse': 'group-label expand' \"\r\n                                (click)=\"changeStatus(category)\">{{ category.categoryName }}</span>\r\n                            <div [style.display]=\"!category.status || category.status === 'open'?'block':'none'\"\r\n                                class=\"py-1 \">\r\n                                <webide-property-item-list [category]=\"category\" [propertyData]=\"propertyData\"\r\n                                    (valueChanged)=\"_valueChanged($event)\" (submitModal)=\"_submitModal($event)\"\r\n                                    (triggerRefreshPanel)=\"refreshPanel()\"></webide-property-item-list>\r\n                            </div>\r\n                        </ng-container>\r\n                    </li>\r\n                </ul>\r\n            </perfect-scrollbar>\r\n\r\n        </div>\r\n    </div>\r\n</div>",
                    styles: [":host{width:inherit;height:inherit}.propertyPanel{width:100%;height:100%;color:rgba(0,0,0,.75);background:#f3f8ff!important;border-color:#d8dbe2!important}.propertyPanel .title{position:relative;background:#d0d8e9!important}.propertyPanel .title>.title-label{line-height:35px;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;padding-left:12px;color:#333}.propertyPanel .title .title-actions{position:absolute;right:10px;top:0;padding-left:35px;height:35px;flex:1;box-sizing:border-box}.propertyPanel .property-grid .group-label{line-height:30px;display:block;padding-left:12px;cursor:pointer;background:#e2e9f6!important;color:#34495e!important}.propertyPanel .panel-body{overflow:auto;height:100%}.propertyPanel .property-grid{list-style:none;margin:0;padding:0}.propertyPanel .action-item{cursor:pointer;display:inline-block;transition:transform 50ms;transition:transform 50ms,-webkit-transform 50ms;position:relative;padding:0}.propertyPanel .action-item .hide-panel{height:35px;display:block;line-height:35px;min-width:28px;color:rgba(66,66,66,.75);background-color:transparent!important;outline:0!important;background:url(\"data:image/svg+xml;charset=utf-8,%3Csvg width='16' height='16' viewBox='0 0 16 16' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M9.784 8L13 11.217 11.215 13 8.001 9.786 4.785 13 3 11.216l3.214-3.215L3 4.785 4.784 3 8 6.216 11.216 3 13 4.785 9.784 8.001z' fill='%23424242'/%3E%3C/svg%3E\") 50% 50% no-repeat}"]
                }] }
    ];
    /** @nocollapse */
    PropertyPanelComponent.ctorParameters = function () { return []; };
    PropertyPanelComponent.propDecorators = {
        width: [{ type: Input }],
        height: [{ type: Input }],
        isPersitOpenState: [{ type: Input }],
        isShowPanel: [{ type: Input }],
        closePropertyPanel: [{ type: Output }],
        propertyConfig: [{ type: Input }],
        propertyData: [{ type: Input }],
        propertyChanged: [{ type: Output }],
        submitModal: [{ type: Output }],
        showCloseBtn: [{ type: Input }],
        panelTitle: [{ type: Input }],
        eastDiv: [{ type: ViewChild, args: ['east',] }],
        itemLists: [{ type: ViewChildren, args: [PropertyItemListComponent,] }]
    };
    return PropertyPanelComponent;
}());
export { PropertyPanelComponent };
if (false) {
    /** @type {?} */
    PropertyPanelComponent.prototype.width;
    /** @type {?} */
    PropertyPanelComponent.prototype.height;
    /** @type {?} */
    PropertyPanelComponent.prototype.isPersitOpenState;
    /** @type {?} */
    PropertyPanelComponent.prototype.isShowPanel;
    /** @type {?} */
    PropertyPanelComponent.prototype.closePropertyPanel;
    /** @type {?} */
    PropertyPanelComponent.prototype.propertyConfig;
    /** @type {?} */
    PropertyPanelComponent.prototype.propertyData;
    /** @type {?} */
    PropertyPanelComponent.prototype.propertyChanged;
    /** @type {?} */
    PropertyPanelComponent.prototype.submitModal;
    /** @type {?} */
    PropertyPanelComponent.prototype.showCloseBtn;
    /** @type {?} */
    PropertyPanelComponent.prototype.panelTitle;
    /**
     * @type {?}
     * @private
     */
    PropertyPanelComponent.prototype.eastDiv;
    /** @type {?} */
    PropertyPanelComponent.prototype.itemLists;
    /** @type {?} */
    PropertyPanelComponent.prototype.isOpen;
}
