// @gsp-wf/wf-process-editor/esm2015/lib/designer/entity/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export { ActionScope } from "./ActionScope";
export { ActivityExecutorParticipant } from "./ActivityExecutorParticipant";
export { ActualParameter } from "./ActualParameter";
export { AssignConfig } from "./AssignConfig";
export { AssignStrategy } from "./AssignStrategy";
export { BizComponent } from "./BizComponent";
export { BizParameter } from "./BizParameter";
export { ComponentParticipant } from "./ComponentParticipant";
export { MatrixParticipant } from "./MatrixParticipant";
export { MessageReceiver } from "./MessageReceiver";
export { OrganizationLevel } from "./OrganizationLevel";
export { OrganizationParticipant } from "./OrganizationParticipant";
export { Participant } from "./Participant";
export { ParticipantInfo } from "./ParticipantInfo";
export { PositionParticipant } from "./PositionParticipant";
export { ProcExtendProperty } from "./ProcExtendProperty";
export { ReportingLineParticipant } from "./ReportingLineParticipant";
export { UserParticipant } from "./UserParticipant";
export { VariablePerformer } from "./VariablePerformer";
