// @edp-pmf/grapheditor/dist/esm/lib/action/Action.js
import { __extends } from "tslib";
import { mxClient, mxEventObject, mxEventSource } from "@edp-pmf/mxgraph-ts";
import { PmfEvents } from "../util/PmfEvents";
import { KEYBOARD_MAP } from "../util/KeyboardMap";
/**
 * 动作类
 */
var Action = /** @class */ (function (_super) {
    __extends(Action, _super);
    function Action(key, editorUi, enabled) {
        var _this = _super.call(this) || this;
        /**
         * 是否启用
         */
        _this.enabled = true;
        /**
         * 是否是切换动作
         */
        _this.toggleAction = false;
        _this.key = key;
        _this.editorUi = editorUi;
        _this.enabled = (enabled != null) ? enabled : true;
        return _this;
    }
    /**
     * 获取启用状态
     */
    Action.prototype.isEnabled = function () {
        return this.enabled;
    };
    /**
     * 获取是否为切换动作
     */
    Action.prototype.isToggleAction = function () {
        return this.toggleAction;
    };
    /**
     * 是否被选中
     */
    Action.prototype.isSelected = function () {
        return false;
    };
    /**
     * 设置启用状态，状态修改时发射状态变更事件
     */
    Action.prototype.setEnabled = function (value) {
        var origin = this.enabled;
        this.enabled = value;
        if (origin != value) {
            this.fireEvent(new mxEventObject(PmfEvents.StateChanged));
        }
    };
    /**
     * 获取快捷键文本
     */
    Action.prototype.getShortcut = function () {
        if (this.shortcutKeyCode == null)
            return null;
        var shortcut = KEYBOARD_MAP[this.shortcutKeyCode];
        if (this.shortcutAltKey) {
            shortcut = "Alt+" + shortcut;
        }
        if (this.shortcutShiftKey) {
            shortcut = "Shift+" + shortcut;
        }
        if (this.shortcutControlKey) {
            shortcut = Action.ctrlKey + "+" + shortcut;
        }
        return shortcut;
    };
    /**
     * 设置是否为切换动作
     */
    Action.prototype.setToggleAction = function (value) {
        this.toggleAction = value;
    };
    /**
     * 执行动作
     */
    Action.prototype.execute = function (evt) {
    };
    Action.ctrlKey = (mxClient.IS_MAC) ? 'Cmd' : 'Ctrl';
    return Action;
}(mxEventSource));
export { Action };
