// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/serializer-context/ubcc-serializer-context.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/serializer-context/ubcc-serializer-context.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DefaultOperationSerializerFactory, DefaultPropertySerializerFactory } from '@ecp-caf/common-structure';
import { UbccParamSerializerFactory } from '../factory/ubcc-param-serializer-factory';
import { UbccDataTypeSerializerFactory } from '../factory/ubcc-data-type-serializer-factory';
import { UbccParamConstraintSerializerFactory } from '../factory/ubcc-param-constraint-serializer-factory';
var UbccParamSerializerContext = /** @class */ (function () {
    function UbccParamSerializerContext() {
    }
    /**
     * @return {?}
     */
    UbccParamSerializerContext.prototype.getDataTypeSerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.dataTypeFactory) {
            this.dataTypeFactory = new UbccDataTypeSerializerFactory();
        }
        return this.dataTypeFactory;
    };
    /**
     * @return {?}
     */
    UbccParamSerializerContext.prototype.getParameterSerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.parameterFactory) {
            this.parameterFactory = new UbccParamSerializerFactory();
        }
        return this.parameterFactory;
    };
    /**
     * @return {?}
     */
    UbccParamSerializerContext.prototype.getOperationSerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.operationFactory) {
            this.operationFactory = new DefaultOperationSerializerFactory();
        }
        return this.operationFactory;
    };
    /**
     * @return {?}
     */
    UbccParamSerializerContext.prototype.getPropertySerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.propertyFactory) {
            this.propertyFactory = new DefaultPropertySerializerFactory();
        }
        return this.propertyFactory;
    };
    /**
     * @return {?}
     */
    UbccParamSerializerContext.prototype.getConstraintSerializerFactory = /**
     * @return {?}
     */
    function () {
        if (!this.constraintFactory) {
            this.constraintFactory = new UbccParamConstraintSerializerFactory();
        }
        return this.constraintFactory;
    };
    return UbccParamSerializerContext;
}());
export { UbccParamSerializerContext };
if (false) {
    /**
     * @type {?}
     * @private
     */
    UbccParamSerializerContext.prototype.dataTypeFactory;
    /**
     * @type {?}
     * @private
     */
    UbccParamSerializerContext.prototype.parameterFactory;
    /**
     * @type {?}
     * @private
     */
    UbccParamSerializerContext.prototype.operationFactory;
    /**
     * @type {?}
     * @private
     */
    UbccParamSerializerContext.prototype.propertyFactory;
    /**
     * @type {?}
     * @private
     */
    UbccParamSerializerContext.prototype.constraintFactory;
}
