// @farris/dynamic-control-group/esm5/lib/components/input/input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from "@angular/core";
import { FormGroup } from '@angular/forms';
var InputComponent = /** @class */ (function () {
    function InputComponent() {
    }
    /**
     * @return {?}
     */
    InputComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.className = this.control.className ? this.control.className : 'col-12 col-md-6 col-xl-3 col-el-2';
    };
    InputComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-input",
                    template: "\n      <div class=\"farris-group-wrap\" [formGroup]=\"form\">\n        <div class=\"form-group farris-form-group\">\n          <label class=\"col-form-label\">\n            <span *ngIf=\"0\" class=\"farris-label-info text-danger\">*</span>\n            <span class=\"farris-label-text\">{{control.name}}</span>\n          </label>\n          <div class=\"farris-input-wrap\">\n            <input-group [formControlName]=\"control.labelCode\" [readonly]=\"control.control.readonly\"></input-group>\n          </div>\n        </div>\n      </div>\n  "
                }] }
    ];
    /** @nocollapse */
    InputComponent.ctorParameters = function () { return []; };
    InputComponent.propDecorators = {
        className: [{ type: HostBinding, args: ['class',] }],
        control: [{ type: Input }],
        form: [{ type: Input }]
    };
    return InputComponent;
}());
export { InputComponent };
if (false) {
    /** @type {?} */
    InputComponent.prototype.className;
    /** @type {?} */
    InputComponent.prototype.control;
    /** @type {?} */
    InputComponent.prototype.form;
}
