// @qdp/ide-cmp/esm2015/lib/service/joint-search/joint-search.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { JointSearchManagerService, LcpService } from '@qdp/common';
export class JointSearchService {
    /**
     * @param {?} jointSearchManager
     * @param {?} lcpService
     */
    constructor(jointSearchManager, lcpService) {
        this.jointSearchManager = jointSearchManager;
        this.lcpService = lcpService;
        this.jointSearchInfo = new Subject();
        this.jointSearchManager.jointsearchInfoChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            this.jointSearchInfo.next(value);
        }));
    }
    /**
     * @param {?} id
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    getjointsearch(id, queryId, queryRelativeUrl) {
        return this.jointSearchManager.getjointsearch(id, queryId, queryRelativeUrl);
    }
    /**
     * @param {?} jointSearchInfo
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    savejointsearch(jointSearchInfo, queryRelativeUrl) {
        return this.jointSearchManager.savejointsearch(jointSearchInfo, queryRelativeUrl);
    }
    /**
     * @param {?} formid
     * @param {?} queryID
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    deletejointsearch(formid, queryID, queryRelativeUrl) {
        return this.jointSearchManager.deletejointsearch(formid, queryID, queryRelativeUrl);
    }
    /**
     * @return {?}
     */
    getLanguageList() {
        return this.jointSearchManager.getLanguageList();
    }
    /**
     * @param {?} params
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    getCol(params, queryRelativeUrl) {
        return this.lcpService.getCol(params, queryRelativeUrl);
    }
}
JointSearchService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
JointSearchService.ctorParameters = () => [
    { type: JointSearchManagerService },
    { type: LcpService }
];
if (false) {
    /** @type {?} */
    JointSearchService.prototype.jointSearchInfo;
    /**
     * @type {?}
     * @private
     */
    JointSearchService.prototype.jointSearchManager;
    /**
     * @type {?}
     * @private
     */
    JointSearchService.prototype.lcpService;
}
