// @ecp-caf/caf-common/esm2015/lib/services/languageservice/language.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { CookieService } from 'ngx-cookie-service';
import { CommonUtil } from '../../utilities/common.utilities';
export class LanguageService {
    /**
     * @param {?} cookieService
     */
    constructor(cookieService) {
        this.cookieService = cookieService;
        this.languageKey = 'languageCode';
        this.curLangugaeKey = 'caf_web_language';
    }
    /**
     * @param {?} languageCode
     * @return {?}
     */
    setLanguageCode(languageCode) {
        localStorage.setItem(this.languageKey, languageCode);
    }
    /**
     * @return {?}
     */
    getLanguageCode() {
        /** @type {?} */
        let curLancode = this.cookieService && this.cookieService.get(this.curLangugaeKey);
        if (!curLancode) {
            curLancode = CommonUtil.getCookieMapFromDoc().get(this.curLangugaeKey) ||
                localStorage.getItem(this.languageKey) ||
                'zh-CHS';
        }
        return curLancode;
    }
}
LanguageService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
LanguageService.ctorParameters = () => [
    { type: CookieService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    LanguageService.prototype.languageKey;
    /**
     * @type {?}
     * @private
     */
    LanguageService.prototype.curLangugaeKey;
    /**
     * @type {?}
     * @private
     */
    LanguageService.prototype.cookieService;
}
