// @gsp-svc/data-entity-adapter/fesm2015/gsp-svc-data-entity-adapter.js
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { Injectable, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * data-entity-service对外提供服务类
 */
class DataEntityService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * \@description 根据Id获取Schema
     * @param {?} id be或vo的ID
     * @return {?} Schema
     */
    getSchemaById(id) {
        return this.http.get("/api/runtime/sys/v1.0/dataentity/schema/id/" + id);
    }
    /**
     * 根据对象获取
     * @param {?} map
     * @return {?}
     */
    getSchemaByObj(map) {
        return this.http.post("/api/runtime/sys/v1.0/dataentity/schema/obj", map);
    }
    /**
     * 获取BE/VO
     * @param {?} id
     * @return {?}
     */
    getBEOrVo(id) {
        return this.http.get("/api/runtime/sys/v1.0/dataentity/schema/bevo/" + id);
    }
}
DataEntityService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DataEntityService.ctorParameters = () => [
    { type: HttpClient }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DataEntityModule {
}
DataEntityModule.decorators = [
    { type: NgModule, args: [{
                declarations: [],
                imports: [
                    HttpClientModule
                ],
                entryComponents: [],
                exports: [],
                providers: [DataEntityService, HttpClient]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { DataEntityModule, DataEntityService };
