// @farris/ui-number-spinner/esm2015/lib/number-only-directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input } from '@angular/core';
import { HostListener } from '@angular/core';
export class NumberOnlyDirective {
    /**
     * @param {?} element
     */
    constructor(element) {
        this.element = element;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onKeyDown(e) {
        /** @type {?} */
        const val = ((/** @type {?} */ (e.target))).value;
        /** @type {?} */
        const key = e.key;
        /** @type {?} */
        const checkKeys = ['.', '-', '%'];
        /** @type {?} */
        const selectionVal = this.getSelectionValue(e.target);
        if (val && checkKeys.indexOf(key) > -1 && val.indexOf(key) > -1 && !(selectionVal && selectionVal.indexOf(key) > -1)) {
            e.preventDefault();
        }
        if ([46, 8, 9, 27, 13, 110, 190, 173].indexOf(e.keyCode) !== -1 ||
            // 允许全选: Ctrl+A
            (e.keyCode === 65 && (e.ctrlKey || e.metaKey)) ||
            // 允许复制: Ctrl+C
            (e.keyCode === 67 && (e.ctrlKey || e.metaKey)) ||
            // 允许粘贴: Ctrl+V
            (e.keyCode === 86 && (e.ctrlKey || e.metaKey)) ||
            // 允许剪切: Ctrl+X
            (e.keyCode === 88 && (e.ctrlKey || e.metaKey)) ||
            // 允许: home（头）, end（尾）, left（左移）, right（右移）
            (e.keyCode >= 35 && e.keyCode <= 39) ||
            e.keyCode === 189 || e.keyCode === 109) {
            // let it happen, don't do anything
            return;
        }
        // 确保数字以外的案件被拒绝执行默认动作
        if ((e.shiftKey || (e.keyCode < 48 || e.keyCode > 57)) && (e.keyCode < 96 || e.keyCode > 105)) {
            e.preventDefault();
        }
        // 判断当前小数精度
        if (val.indexOf('.') > -1 && this.precision) {
            /** @type {?} */
            const curPos = ((/** @type {?} */ (e.target))).selectionStart;
            /** @type {?} */
            const selectionEnd = ((/** @type {?} */ (e.target))).selectionEnd;
            /** @type {?} */
            const _precision = val.split('.')[1].length;
            /** @type {?} */
            const dotPos = val.indexOf('.');
            if (curPos > dotPos) {
                /** @type {?} */
                const selectionText = val.substring(curPos, selectionEnd);
                if (_precision - selectionText.length >= Number(this.precision)) {
                    e.preventDefault();
                    return false;
                }
            }
        }
    }
    /**
     * @private
     * @param {?} target
     * @return {?}
     */
    getSelectionValue(target) {
        /** @type {?} */
        const start = ((/** @type {?} */ (target))).selectionStart;
        /** @type {?} */
        const end = ((/** @type {?} */ (target))).selectionEnd;
        return target.value.substring(start, end);
    }
}
NumberOnlyDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line: directive-selector
                selector: '[number-only]'
            },] }
];
/** @nocollapse */
NumberOnlyDirective.ctorParameters = () => [
    { type: ElementRef }
];
NumberOnlyDirective.propDecorators = {
    precision: [{ type: Input }],
    onKeyDown: [{ type: HostListener, args: ['keydown', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    NumberOnlyDirective.prototype.precision;
    /**
     * @type {?}
     * @private
     */
    NumberOnlyDirective.prototype.element;
}
