// @gsp-lcm/bo-rt-service/fesm5/gsp-lcm-bo-rt-service.js
import { of } from 'rxjs/observable/of';
import { Injectable, NgModule } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/business-object-rt.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var PROJECTS_DIRECTORY = '/api/dev/main/v1.0/business-object-trees';
/** @type {?} */
var url = '/api/runtime/sys/v1.0/business-objects/';
var BusinessObjectRtService = /** @class */ (function () {
    function BusinessObjectRtService(http) {
        this.http = http;
    }
    /**
     * @return {?}
     */
    BusinessObjectRtService.prototype.getBusinessObjectTree = /**
     * @return {?}
     */
    function () {
        return this.http.get(PROJECTS_DIRECTORY).pipe();
    };
    /**
     * @param {?} boId
     * @return {?}
     */
    BusinessObjectRtService.prototype.getBusinessObject = /**
     * @param {?} boId
     * @return {?}
     */
    function (boId) {
        /** @type {?} */
        var getUrl = url + boId;
        return this.http.get(getUrl).pipe();
    };
    /**
     * @return {?}
     */
    BusinessObjectRtService.prototype.getBusinessObjects = /**
     * @return {?}
     */
    function () {
        return this.http.get(url).pipe();
    };
    /**
   * 保存
   */
    /**
     * 保存
     * @param {?} postData
     * @return {?}
     */
    BusinessObjectRtService.prototype.save = /**
     * 保存
     * @param {?} postData
     * @return {?}
     */
    function (postData) {
        return this.http.post(url, postData).pipe();
    };
    /**
     * @param {?} postData
     * @return {?}
     */
    BusinessObjectRtService.prototype.saveChanges = /**
     * @param {?} postData
     * @return {?}
     */
    function (postData) {
        return this.http.put(url, postData).pipe();
    };
    /**
       * 删除并重新加载
       */
    /**
     * 删除并重新加载
     * @param {?} dataId
     * @return {?}
     */
    BusinessObjectRtService.prototype.remove = /**
     * 删除并重新加载
     * @param {?} dataId
     * @return {?}
     */
    function (dataId) {
        /** @type {?} */
        var deleteUrl = url + dataId;
        return this.http.delete(deleteUrl, null).pipe();
    };
    /**
     * @param {?} boId
     * @return {?}
     */
    BusinessObjectRtService.prototype.getDevBasicBOInfo = /**
     * @param {?} boId
     * @return {?}
     */
    function (boId) {
        /** @type {?} */
        var devUrl = url + 'dev-basic-boinfos/';
        /** @type {?} */
        var getUrl = devUrl + boId;
        return this.http.get(getUrl).pipe();
    };
    /**
     * @param {?} parentId
     * @return {?}
     */
    BusinessObjectRtService.prototype.getBOListByParentId = /**
     * @param {?} parentId
     * @return {?}
     */
    function (parentId) {
        /** @type {?} */
        var getUrl = url + '?parentID=' + parentId;
        return this.http.get(getUrl).pipe();
    };
    /**
     * @private
     * @template T
     * @param {?=} operation
     * @param {?=} result
     * @return {?}
     */
    BusinessObjectRtService.prototype.handleError = /**
     * @private
     * @template T
     * @param {?=} operation
     * @param {?=} result
     * @return {?}
     */
    function (operation, result) {
        if (operation === void 0) { operation = 'operation'; }
        return (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            console.error(error); // log to console instead
            return of((/** @type {?} */ (result)));
        });
    };
    BusinessObjectRtService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    BusinessObjectRtService.ctorParameters = function () { return [
        { type: HttpService }
    ]; };
    return BusinessObjectRtService;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/bo-rt-service.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BORtServiceModule = /** @class */ (function () {
    function BORtServiceModule() {
    }
    BORtServiceModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [
                        CommonModule
                    ],
                    exports: [],
                    providers: [HttpService, BusinessObjectRtService]
                },] }
    ];
    return BORtServiceModule;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: gsp-lcm-bo-rt-service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { BusinessObjectRtService, BORtServiceModule };
