// @farris/ui-sidebar/esm5/lib/sidebar.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
var FarrisSidebarService = /** @class */ (function () {
    function FarrisSidebarService() {
        this.isOpenSubject = new Subject();
    }
    /**
     * @param {?} message
     * @return {?}
     */
    FarrisSidebarService.prototype.sendIsOpen = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
        this.isOpenSubject.next(message);
    };
    /**
     * @return {?}
     */
    FarrisSidebarService.prototype.getIsOpen = /**
     * @return {?}
     */
    function () {
        return this.isOpenSubject.asObservable();
    };
    FarrisSidebarService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    FarrisSidebarService.ctorParameters = function () { return []; };
    /** @nocollapse */ FarrisSidebarService.ngInjectableDef = i0.defineInjectable({ factory: function FarrisSidebarService_Factory() { return new FarrisSidebarService(); }, token: FarrisSidebarService, providedIn: "root" });
    return FarrisSidebarService;
}());
export { FarrisSidebarService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FarrisSidebarService.prototype.isOpenSubject;
}
