// @gsp-aif/runtime-api/esm2015/lib/aif-pull-select-creation-rule/aif-pull-select-creation-rule.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from "@angular/core";
import { AifPullSelectCreationRuleInternalService } from "./service/aif-pull-select-creation-rule.internal.service";
import { CommonModule } from "@angular/common";
import { HttpClientModule } from "@angular/common/http";
import { ButtonsModule } from "@progress/kendo-angular-buttons";
import { DropDownsModule } from "@progress/kendo-angular-dropdowns";
import { SplitterModule } from "@progress/kendo-angular-layout";
import { AifPullSelectCreationRuleComponent } from "./aif-pull-select-creation-rule/aif-pull-select-creation-rule.component";
import { HttpService } from "@ecp-caf/caf-common";
import { FormsModule, ReactiveFormsModule } from "@angular/forms";
import { ModalModule, BsModalService } from '@farris/ui-modal';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { AifPullSelectCreationRuleService } from './service/aif-pull-select-creation-rule.service';
import { AifPullSelectCreationRuleModuleConfig } from '../entity/aif-pull-select-creation-rule-module-config';
const ɵ0 = {};
export class AifPullSelectCreationRuleModule {
    /**
     * @param {?=} config
     * @return {?}
     */
    static forRoot(config) {
        return {
            "ngModule": AifPullSelectCreationRuleModule,
            "providers": [
                { provide: AifPullSelectCreationRuleModuleConfig, useValue: config || {} }
            ]
        };
    }
}
AifPullSelectCreationRuleModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                    HttpClientModule,
                    ButtonsModule,
                    DropDownsModule,
                    SplitterModule,
                    ModalModule,
                    FarrisDialogModule,
                ],
                providers: [
                    HttpService,
                    BsModalService,
                    { provide: AifPullSelectCreationRuleModuleConfig, useValue: ɵ0 },
                    AifPullSelectCreationRuleInternalService,
                    AifPullSelectCreationRuleService
                ],
                declarations: [
                    AifPullSelectCreationRuleComponent
                ],
                bootstrap: [
                    AifPullSelectCreationRuleComponent
                ],
                entryComponents: [
                    AifPullSelectCreationRuleComponent
                ],
                exports: []
            },] }
];
export { ɵ0 };
