// @farris/ui-property-panel/esm5/lib/entity/property-entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var PropertyType = {
    string: 'string',
    boolean: 'boolean',
    number: 'number',
    select: 'select',
    boolOrExp: 'boolOrExp',
    multiSelect: 'multiSelect',
    date: 'date',
    datetime: 'datetime',
    modal: 'modal',
    cascade: 'cascade',
};
export { PropertyType };
var ElementPropertyConfig = /** @class */ (function () {
    function ElementPropertyConfig() {
        this.hide = false;
        this.hideTitle = false;
        // 是否启用级联特性
        this.enableCascade = false;
    }
    return ElementPropertyConfig;
}());
export { ElementPropertyConfig };
if (false) {
    /** @type {?} */
    ElementPropertyConfig.prototype.categoryId;
    /** @type {?} */
    ElementPropertyConfig.prototype.categoryName;
    /** @type {?} */
    ElementPropertyConfig.prototype.hide;
    /** @type {?} */
    ElementPropertyConfig.prototype.hideTitle;
    /** @type {?} */
    ElementPropertyConfig.prototype.properties;
    /** @type {?} */
    ElementPropertyConfig.prototype.enableCascade;
    /** @type {?} */
    ElementPropertyConfig.prototype.propertyData;
    /** @type {?} */
    ElementPropertyConfig.prototype.parentPropertyID;
    /** @type {?} */
    ElementPropertyConfig.prototype.setPropertyRelates;
}
// 属性实体
var 
// 属性实体
PropertyEntity = /** @class */ (function () {
    function PropertyEntity() {
        // 属性的默认值
        this.readonly = false; // 是否只读
        // 是否只读
        this.visible = true; // 是否可见
        // 数字类型属性的小数位数
        // 属性改变后刷新整个面板：用于更改其他分类下的属性
        this.refreshPanelAfterChanged = false;
        // 下拉多选类型：属性值的类型：string(多值以逗号分隔)/array(多值组装成数组)
        this.multiSelectDataType = 'string';
        // 级联属性配置
        this.isExpand = false; // 级联属性是否默认收起
        // 级联属性是否默认收起
        this.hideCascadeTitle = false; // 是否隐藏级联属性的头部
        // 属性值转换器,
        this.showClearButton = false; // 是否展示清除图标
    }
    // 打开模态框前的方法，一般用于校验逻辑，返回值中result=true，则进一步打开模态框，result=false则提示message内容，并不再打开模态框。
    // 是否展示清除图标
    // 打开模态框前的方法，一般用于校验逻辑，返回值中result=true，则进一步打开模态框，result=false则提示message内容，并不再打开模态框。
    /**
     * @return {?}
     */
    PropertyEntity.prototype.beforeOpenModal = 
    // 是否展示清除图标
    // 打开模态框前的方法，一般用于校验逻辑，返回值中result=true，则进一步打开模态框，result=false则提示message内容，并不再打开模态框。
    /**
     * @return {?}
     */
    function () {
        return new BeforeOpenModalResult();
    };
    return PropertyEntity;
}());
// 属性实体
export { PropertyEntity };
if (false) {
    /** @type {?} */
    PropertyEntity.prototype.propertyID;
    /** @type {?} */
    PropertyEntity.prototype.propertyName;
    /** @type {?} */
    PropertyEntity.prototype.propertyType;
    /** @type {?} */
    PropertyEntity.prototype.description;
    /** @type {?} */
    PropertyEntity.prototype.defaultValue;
    /** @type {?} */
    PropertyEntity.prototype.readonly;
    /** @type {?} */
    PropertyEntity.prototype.visible;
    /** @type {?} */
    PropertyEntity.prototype.iterator;
    /** @type {?} */
    PropertyEntity.prototype.min;
    /** @type {?} */
    PropertyEntity.prototype.max;
    /** @type {?} */
    PropertyEntity.prototype.numberFormat;
    /** @type {?} */
    PropertyEntity.prototype.decimals;
    /** @type {?} */
    PropertyEntity.prototype.refreshPanelAfterChanged;
    /** @type {?} */
    PropertyEntity.prototype.multiSelectDataType;
    /**
     * 级联属性
     * @type {?}
     */
    PropertyEntity.prototype.cascadeConfig;
    /** @type {?} */
    PropertyEntity.prototype.isExpand;
    /** @type {?} */
    PropertyEntity.prototype.hideCascadeTitle;
    /** @type {?} */
    PropertyEntity.prototype.cascadeConverter;
    /**
     * 模态框属性配置
     * @type {?}
     */
    PropertyEntity.prototype.editor;
    /** @type {?} */
    PropertyEntity.prototype.editorParams;
    /** @type {?} */
    PropertyEntity.prototype.converter;
    /** @type {?} */
    PropertyEntity.prototype.showClearButton;
    /* Skipping unhandled member: [propName: string]: any;*/
}
/**
 * @record
 */
export function TypeConverter() { }
if (false) {
    /**
     * @param {?} data
     * @param {?=} params
     * @return {?}
     */
    TypeConverter.prototype.convertTo = function (data, params) { };
}
/**
 * @record
 */
export function KeyMap() { }
if (false) {
    /** @type {?} */
    KeyMap.prototype.key;
    /** @type {?} */
    KeyMap.prototype.value;
}
var BeforeOpenModalResult = /** @class */ (function () {
    function BeforeOpenModalResult() {
    }
    return BeforeOpenModalResult;
}());
export { BeforeOpenModalResult };
if (false) {
    /** @type {?} */
    BeforeOpenModalResult.prototype.result;
    /** @type {?} */
    BeforeOpenModalResult.prototype.message;
}
