// @gspwidget/widget-devkit/esm2015/lib/filter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import add from "date-fns/add";
import sub from 'date-fns/sub';
import { PropertyTypes } from "../types";
/**
 * æ²¡å‘ç›´æŽ¥ç”¨ [] ä½œ key ç›´æŽ¥å»º literal å¯¹è±¡ï¼Œaot ç¼–è¯‘ä¼šæŠ¥é”™
 * @param {?} key
 * @param {?} value
 * @return {?}
 */
export function getObj(key, value) {
    /** @type {?} */
    const obj = {};
    obj[key] = value;
    return obj;
}
/**
 * ç­›é€‰æè¿°
 */
export class FSLastTimeItem {
    constructor() {
        this.num = 1;
        this.unit = 'day';
    }
    /**
     * ä¼ å…¥ä¸€ä¸ª 'æœ€è¿‘næŸ'ï¼Œè¿”å›žå¯¹åº”çš„æ—¶é—´èŒƒå›´æ•°ç»„
     * @param {?} lastTimeItem
     * @return {?}
     */
    static GetDates(lastTimeItem) {
        return [
            sub(new Date(), getObj(lastTimeItem.unit + 's', lastTimeItem.num)),
            new Date()
        ];
    }
}
if (false) {
    /** @type {?} */
    FSLastTimeItem.prototype.num;
    /** @type {?} */
    FSLastTimeItem.prototype.unit;
    /** @type {?} */
    FSLastTimeItem.prototype.isDefault;
}
export class FSRangeDefault {
    constructor() {
        this.beforeNum = 1;
        this.beforeUnit = 'month';
        this.afterNum = 0;
        this.afterUnit = 'day';
    }
    /**
     * @param {?} rangeDefault
     * @return {?}
     */
    static GetDates(rangeDefault) {
        return [
            sub(new Date(), getObj(rangeDefault.beforeUnit + 's', rangeDefault.afterNum)),
            add(new Date(), getObj(rangeDefault.beforeUnit + 's', rangeDefault.afterNum)),
        ];
    }
}
if (false) {
    /** @type {?} */
    FSRangeDefault.prototype.beforeNum;
    /** @type {?} */
    FSRangeDefault.prototype.beforeUnit;
    /** @type {?} */
    FSRangeDefault.prototype.afterNum;
    /** @type {?} */
    FSRangeDefault.prototype.afterUnit;
}
export class FSDate {
    constructor() {
        this.type = 'single';
        this.singleUnit = 'day';
        this.singleDefaultType = 'current';
        this.singleDefaultSpecified = new Date();
        this.rangeDefault = new FSRangeDefault();
        this.lastTimeItems = [];
        this.lastTimeDisplayType = 'flat';
    }
}
if (false) {
    /** @type {?} */
    FSDate.prototype.type;
    /** @type {?} */
    FSDate.prototype.singleUnit;
    /** @type {?} */
    FSDate.prototype.singleDefaultType;
    /** @type {?} */
    FSDate.prototype.singleDefaultSpecified;
    /** @type {?} */
    FSDate.prototype.rangeDefault;
    /** @type {?} */
    FSDate.prototype.lastTimeItems;
    /** @type {?} */
    FSDate.prototype.lastTimeDisplayType;
}
export class FSEnum {
    constructor() {
        this.multi = false;
        /**
         * æš‚æ—¶å®šä¸ºï¼Œä»…æœ‰ single æ—¶å¯ä»¥é€‰æ‹© flat
         */
        this.displayType = 'select';
        this.source = 'manual';
        /**
         * æ˜¯å¦å¯æ¸…ç©ºä¸é€‰ã€‚ä»…æœ‰ select ä¸” single æ‰æœ‰æ­¤é€‰é¡¹
         */
        this.allowEmpty = false;
        /**
         * é€šè¿‡ä¸Šé¢ä¸¤ä¸ªå­—æ®µå–å‡º items å­˜åœ¨ä¸‹é¢
         */
        this.itemsFromData = [];
        this.itemsManual = [];
        this.multiDefault = [];
    }
    // æš‚æ—¶ä¸åš
    // sourceDatasetId: string
    // sourceDatasetLabelField?: DataField
    // sourceDatasetValueField?: DataField
    /**
     * @param {?} fsEnum
     * @return {?}
     */
    static GetItems(fsEnum) {
        const { source, itemsFromData, itemsManual } = fsEnum;
        if (source === 'dataField') {
            // if (!itemsValueField || !intactData) return []
            // const tempItems = intactData.map(dataItem => ({
            //   label: dataItem[itemsLabelField || itemsValueField],
            //   value: dataItem[itemsValueField]
            // }))
            // return ArrUtil.Unique(tempItems)
            return itemsFromData;
        }
        else if (source === 'manual') {
            return itemsManual;
        }
    }
    /**
     * @param {?} fsEnum
     * @return {?}
     */
    static GetItemValues(fsEnum) {
        /**
         * @param {?} item
         * @return {?}
         */
        function mapCb(item) { return item.value; }
        return FSEnum.GetItems(fsEnum).map(mapCb);
    }
}
if (false) {
    /** @type {?} */
    FSEnum.prototype.multi;
    /**
     * æš‚æ—¶å®šä¸ºï¼Œä»…æœ‰ single æ—¶å¯ä»¥é€‰æ‹© flat
     * @type {?}
     */
    FSEnum.prototype.displayType;
    /** @type {?} */
    FSEnum.prototype.source;
    /**
     * æ˜¯å¦å¯æ¸…ç©ºä¸é€‰ã€‚ä»…æœ‰ select ä¸” single æ‰æœ‰æ­¤é€‰é¡¹
     * @type {?}
     */
    FSEnum.prototype.allowEmpty;
    /** @type {?} */
    FSEnum.prototype.itemsLabelField;
    /** @type {?} */
    FSEnum.prototype.itemsValueField;
    /**
     * é€šè¿‡ä¸Šé¢ä¸¤ä¸ªå­—æ®µå–å‡º items å­˜åœ¨ä¸‹é¢
     * @type {?}
     */
    FSEnum.prototype.itemsFromData;
    /** @type {?} */
    FSEnum.prototype.itemsManual;
    /** @type {?} */
    FSEnum.prototype.multiDefault;
    /** @type {?} */
    FSEnum.prototype.singleDefault;
}
export class FSEnumItem {
}
if (false) {
    /** @type {?} */
    FSEnumItem.prototype.label;
    /** @type {?} */
    FSEnumItem.prototype.value;
}
export class FilterSchema {
    constructor() {
        /**
         * è¿‡æ»¤åç§°ï¼Œåœ¨ç­›é€‰æŽ§ä»¶ä¸Šæ˜¾ç¤ºï¼Œå¯é€‰
         */
        this.i18nName = {};
        this.showName = false;
        /**
         * è¿‡æ»¤å¤§ç±»
         */
        this.type = 'date';
        /**
         * æ¯ç§ç±»åž‹çš„è¿‡æ»¤é…ç½®
         */
        this.date = new FSDate();
        this.enum = new FSEnum();
        this.number = {
            default: 0
        };
        this.text = {
            default: ''
        };
    }
    /**
     * èŽ·å– FS çš„é»˜è®¤å€¼
     * ðŸ™… ä¸å¯ä»¥ä¾èµ– dataï¼šéƒ¨ä»¶ç¬¬ä¸€æ¬¡æ ¹æ®é»˜è®¤å€¼ç­›é€‰æ—¶ï¼Œè¿˜æ²¡æœ‰ data å‘¢ã€‚
     * @param {?} fs
     * @return {?}
     */
    static getDefaultValue(fs) {
        switch (fs.type) {
            case 'date':
                if (fs.date.type === 'single') {
                    if (fs.date.singleDefaultType === 'current') {
                        return new Date();
                    }
                    else {
                        return fs.date.singleDefaultSpecified;
                    }
                }
                else if (fs.date.type === 'range') {
                    return FSRangeDefault.GetDates(fs.date.rangeDefault);
                }
                else if (fs.date.type === 'lastTimeEnum') {
                    return FSLastTimeItem.GetDates(fs.date.lastTimeItems[0]);
                }
                break;
            case 'enum':
                const { multi, singleDefault, multiDefault } = fs.enum;
                return multi ? multiDefault : singleDefault;
            case 'number':
                return fs.number.default;
                break;
            case 'text':
                return fs.text.default;
                break;
        }
    }
}
if (false) {
    /**
     * è¿‡æ»¤åç§°ï¼Œåœ¨ç­›é€‰æŽ§ä»¶ä¸Šæ˜¾ç¤ºï¼Œå¯é€‰
     * @type {?}
     */
    FilterSchema.prototype.i18nName;
    /** @type {?} */
    FilterSchema.prototype.showName;
    /**
     * è¿‡æ»¤å¤§ç±»
     * @type {?}
     */
    FilterSchema.prototype.type;
    /**
     * æ¯ç§ç±»åž‹çš„è¿‡æ»¤é…ç½®
     * @type {?}
     */
    FilterSchema.prototype.date;
    /** @type {?} */
    FilterSchema.prototype.enum;
    /** @type {?} */
    FilterSchema.prototype.help;
    /** @type {?} */
    FilterSchema.prototype.number;
    /** @type {?} */
    FilterSchema.prototype.text;
}
/**
 * @param {?} fs
 * @return {?}
 */
export function ifFsDateRange(fs) {
    if (!fs)
        return;
    return fs.type === 'date' && fs.date.type === 'range';
}
/**
 * @param {?} fs
 * @return {?}
 */
export function isFsFlat(fs) {
    if (!fs)
        return;
    return fs.type === 'date' && fs.date.type === 'lastTimeEnum' && fs.date.lastTimeDisplayType === 'flat'
        || fs.type === 'enum' && !fs.enum.multi && fs.enum.displayType === 'flat';
}
/**
 * å®˜æ–¹é¢„åˆ¶çš„ç­›é€‰ property
 * @type {?}
 */
export const FILTER_PROP_NAME = 'filter';
/** @type {?} */
export const frontendFilterProp = {
    name: FILTER_PROP_NAME,
    displayName: 'ç­›é€‰',
    isArray: true,
    type: PropertyTypes.Object,
    objectOption: {
        objPropertyOptions: [{
                name: 'fs',
                type: PropertyTypes.Filter,
            }, {
                name: 'dpId',
                type: PropertyTypes.DataParam,
            }, {
                name: 'dpId2',
                type: PropertyTypes.DataParam,
                ifShow: (/**
                 * @param {?} pv
                 * @return {?}
                 */
                pv => {
                    return pv.fs && ifFsDateRange(pv.fs);
                })
            }]
    }
};
/**
 * @record
 */
export function FilterConfig() { }
if (false) {
    /** @type {?} */
    FilterConfig.prototype.fs;
    /** @type {?} */
    FilterConfig.prototype.dpId;
    /** @type {?} */
    FilterConfig.prototype.dpId2;
}
