// @gsp-wf/wf-process-editor/esm5/lib/views/participant/participant.info.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { AssignStrategy } from "../assign-strategy/assign-strategy";
var ParticipantInfo = /** @class */ (function () {
    function ParticipantInfo() {
        this.assignStrategy = new AssignStrategy();
        this.humanPerformerList = new Array();
        this.potentialOwnerList = new Array();
        this.isAllScope = false;
    }
    return ParticipantInfo;
}());
export { ParticipantInfo };
if (false) {
    /** @type {?} */
    ParticipantInfo.prototype.assignStrategy;
    /** @type {?} */
    ParticipantInfo.prototype.humanPerformerList;
    /** @type {?} */
    ParticipantInfo.prototype.potentialOwnerList;
    /** @type {?} */
    ParticipantInfo.prototype.isAllScope;
}
var Participant = /** @class */ (function () {
    function Participant() {
    }
    return Participant;
}());
export { Participant };
if (false) {
    /** @type {?} */
    Participant.prototype.id;
    /** @type {?} */
    Participant.prototype.type;
    /** @type {?} */
    Participant.prototype.condition;
    /** @type {?} */
    Participant.prototype.order;
}
var UserParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(UserParticipant, _super);
    function UserParticipant() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return UserParticipant;
}(Participant));
export { UserParticipant };
if (false) {
    /** @type {?} */
    UserParticipant.prototype.performerId;
    /** @type {?} */
    UserParticipant.prototype.performerName;
    /** @type {?} */
    UserParticipant.prototype.orgId;
    /** @type {?} */
    UserParticipant.prototype.orgName;
    /** @type {?} */
    UserParticipant.prototype.orgType;
    /** @type {?} */
    UserParticipant.prototype.tag;
}
var VariableParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(VariableParticipant, _super);
    function VariableParticipant() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return VariableParticipant;
}(UserParticipant));
export { VariableParticipant };
if (false) {
    /** @type {?} */
    VariableParticipant.prototype.name;
    /** @type {?} */
    VariableParticipant.prototype.varValue;
    /** @type {?} */
    VariableParticipant.prototype.varType;
    /** @type {?} */
    VariableParticipant.prototype.relation;
}
var PositionParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(PositionParticipant, _super);
    function PositionParticipant() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return PositionParticipant;
}(UserParticipant));
export { PositionParticipant };
if (false) {
    /** @type {?} */
    PositionParticipant.prototype.positionType;
    /** @type {?} */
    PositionParticipant.prototype.isRequired;
}
var ComponentParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(ComponentParticipant, _super);
    function ComponentParticipant() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ComponentParticipant;
}(Participant));
export { ComponentParticipant };
if (false) {
    /** @type {?} */
    ComponentParticipant.prototype.name;
    /** @type {?} */
    ComponentParticipant.prototype.metadataId;
    /** @type {?} */
    ComponentParticipant.prototype.methodCode;
    /** @type {?} */
    ComponentParticipant.prototype.methodName;
    /** @type {?} */
    ComponentParticipant.prototype.actualParameters;
}
var ReportingLineParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(ReportingLineParticipant, _super);
    function ReportingLineParticipant() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ReportingLineParticipant;
}(UserParticipant));
export { ReportingLineParticipant };
if (false) {
    /** @type {?} */
    ReportingLineParticipant.prototype.reportingLineDefId;
    /** @type {?} */
    ReportingLineParticipant.prototype.reportingLineDefName;
    /** @type {?} */
    ReportingLineParticipant.prototype.refOrgId;
    /** @type {?} */
    ReportingLineParticipant.prototype.refOrgName;
    /** @type {?} */
    ReportingLineParticipant.prototype.refUserId;
    /** @type {?} */
    ReportingLineParticipant.prototype.refEmployeeId;
    /** @type {?} */
    ReportingLineParticipant.prototype.dimension;
    /** @type {?} */
    ReportingLineParticipant.prototype.reportingReferObject;
}
var MatrixParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(MatrixParticipant, _super);
    function MatrixParticipant() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return MatrixParticipant;
}(UserParticipant));
export { MatrixParticipant };
if (false) {
    /** @type {?} */
    MatrixParticipant.prototype.matrixDefId;
    /** @type {?} */
    MatrixParticipant.prototype.matrixDefName;
    /** @type {?} */
    MatrixParticipant.prototype.conditionField1;
    /** @type {?} */
    MatrixParticipant.prototype.conditionFieldText1;
    /** @type {?} */
    MatrixParticipant.prototype.conditionField2;
    /** @type {?} */
    MatrixParticipant.prototype.conditionFieldText2;
    /** @type {?} */
    MatrixParticipant.prototype.conditionField3;
    /** @type {?} */
    MatrixParticipant.prototype.conditionFieldText3;
    /** @type {?} */
    MatrixParticipant.prototype.dataFieldAlias;
}
var ActivityExecutorParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(ActivityExecutorParticipant, _super);
    function ActivityExecutorParticipant() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ActivityExecutorParticipant;
}(UserParticipant));
export { ActivityExecutorParticipant };
if (false) {
    /** @type {?} */
    ActivityExecutorParticipant.prototype.activityDefId;
    /** @type {?} */
    ActivityExecutorParticipant.prototype.activityDefName;
}
var OrganizationParticipant = /** @class */ (function (_super) {
    tslib_1.__extends(OrganizationParticipant, _super);
    function OrganizationParticipant() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return OrganizationParticipant;
}(Participant));
export { OrganizationParticipant };
if (false) {
    /** @type {?} */
    OrganizationParticipant.prototype.name;
    /** @type {?} */
    OrganizationParticipant.prototype.actionScope;
    /** @type {?} */
    OrganizationParticipant.prototype.organizationLevel;
    /** @type {?} */
    OrganizationParticipant.prototype.organizationId;
    /** @type {?} */
    OrganizationParticipant.prototype.organizationCode;
    /** @type {?} */
    OrganizationParticipant.prototype.organizationName;
    /** @type {?} */
    OrganizationParticipant.prototype.includeChildren;
    /** @type {?} */
    OrganizationParticipant.prototype.parentsUntil;
    /** @type {?} */
    OrganizationParticipant.prototype.tag;
}
/** @enum {string} */
var ActionScope = {
    /**
     * 指派
     */
    AssignUser: 'AssignUser',
    /**
     * 移交
     */
    Transfer: 'Transfer',
    /**
     * 征询
     */
    AddSignFront: 'AddSignFront',
    /**
     * 加签
     */
    AddSignBehind: 'AddSignBehind',
    /**
     * 知会
     */
    Notify: 'Notify',
};
export { ActionScope };
/** @enum {string} */
var OrganizationLevel = {
    /**
     * 部门
     */
    DEPARTMENT: 'DEPARTMENT',
    /**
     * 单位
     */
    UNIT: 'UNIT',
    /**
    * 默认值
    */
    DEFAULT: 'DEFAULT',
    VARORGANIZATION: 'VARORGANIZATION',
};
export { OrganizationLevel };
var ActualParameter = /** @class */ (function () {
    function ActualParameter() {
    }
    return ActualParameter;
}());
export { ActualParameter };
if (false) {
    /** @type {?} */
    ActualParameter.prototype.code;
    /** @type {?} */
    ActualParameter.prototype.value;
    /** @type {?} */
    ActualParameter.prototype.valueText;
    /** @type {?} */
    ActualParameter.prototype.tag;
    /** @type {?} */
    ActualParameter.prototype.fetchValueMode;
    /** @type {?} */
    ActualParameter.prototype.fetchValueSetting;
}
