// @farris/mobile-devkit/esm2015/lib/repository/data_change_history.js
import { DataChangeType } from '../core';
class DataChangeHistory {
    constructor() {
        this.history = [];
        this.cacheData = [];
    }
    addChange(dataChange) {
        const changeType = DataChangeType[dataChange.changeType];
        this[`on${changeType}Data`](dataChange);
    }
    addChanges(dataChange) {
        dataChange.forEach(change => this.addChange(change));
    }
    clear() {
        this.history.splice(0, this.history.length);
    }
    clearByIds(ids) {
        this.history = this.history.filter(item => {
            if (item.fpath && item.fpath !== '/' && item.fpath.includes('/')) {
                for (const id of ids) {
                    const include = item.fpath.split('/').includes(id);
                    return !include;
                }
            }
            else {
                return !ids.includes(item.dataId);
            }
        });
    }
    isChanged() {
        return this.history.length > 0;
    }
    clearCache() {
        this.cacheData.splice(0, this.cacheData.length);
    }
    getCacheData() {
        return this.cacheData;
    }
    onAddData(dataChange) {
        this.history.push(dataChange);
    }
    onEditData(dataChange) {
        this.cacheData.push(dataChange);
    }
    onAppendData(dataChange) {
        this.history.push(dataChange);
        this.cacheData.push(dataChange);
    }
    onDeleteData(dataChange) {
        const index = this.history.findIndex(item => item.dataId === dataChange.dataId && item.changeType === DataChangeType.Add);
        if (index >= 0) {
            this.history.splice(index, 1);
        }
        else {
            this.history.push(dataChange);
        }
    }
}
export { DataChangeHistory };
