// @gsp-wf/wf-process-model/esm5/lib/model/extension/assignment-strategy/AssignPerformerConfig.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ExtensionElement, BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
var AssignPerformerConfig = /** @class */ (function (_super) {
    tslib_1.__extends(AssignPerformerConfig, _super);
    function AssignPerformerConfig(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = AssignPerformerConfig.Clr_Type_ID;
        _this.MultiSelect = true;
        _this.Necessary = false;
        _this.AllowGlobalAssign = false;
        _this.Operator = 'currentUser';
        return _this;
    }
    /**
     * @return {?}
     */
    AssignPerformerConfig.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new AssignPerformerConfig(this.Model);
    };
    /**
     * @protected
     * @return {?}
     */
    AssignPerformerConfig.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    /**
     * @return {?}
     */
    AssignPerformerConfig.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "Operator", this.Operator);
        BpmnModelHelper.SetAttributeValue(obj, "MultiSelect", this.MultiSelect);
        BpmnModelHelper.SetAttributeValue(obj, "Necessary", this.Necessary);
        BpmnModelHelper.SetAttributeValue(obj, "AllowGlobalAssign", this.AllowGlobalAssign);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AssignPerformerConfig.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Operator = obj.Operator;
        this.MultiSelect = obj.MultiSelect;
        this.Necessary = obj.Necessary;
        this.AllowGlobalAssign = obj.AllowGlobalAssign;
    };
    /**
     * @return {?}
     */
    AssignPerformerConfig.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var assignPerformerConfig = new AssignPerformerConfig(this.Model);
        assignPerformerConfig.Operator = this.Operator;
        assignPerformerConfig.MultiSelect = this.MultiSelect;
        assignPerformerConfig.Necessary = this.Necessary;
        assignPerformerConfig.AllowGlobalAssign = this.AllowGlobalAssign;
        return assignPerformerConfig;
    };
    AssignPerformerConfig.Clr_Type_ID = 'wf.AssignPerformerConfig';
    return AssignPerformerConfig;
}(ExtensionElement));
export { AssignPerformerConfig };
if (false) {
    /** @type {?} */
    AssignPerformerConfig.Clr_Type_ID;
    /** @type {?} */
    AssignPerformerConfig.prototype.ClrTypeID;
    /** @type {?} */
    AssignPerformerConfig.prototype.Operator;
    /** @type {?} */
    AssignPerformerConfig.prototype.MultiSelect;
    /** @type {?} */
    AssignPerformerConfig.prototype.Necessary;
    /** @type {?} */
    AssignPerformerConfig.prototype.AllowGlobalAssign;
}
BpmnModelHelper.RegisterElementType(new AssignPerformerConfig(new BpmnModel()));
