// @gsp-wf/wf-process-model/esm5/lib/model/extension/assignment-strategy/AssignmentStrategyDef.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper, ExtensionElement } from "@edp-pmf/bpmn-model";
var AssignmentStrategyDef = /** @class */ (function (_super) {
    tslib_1.__extends(AssignmentStrategyDef, _super);
    //undo操作临时clone方法
    function AssignmentStrategyDef(model, id, description, expandGroups) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "wf.AssignmentStrategy";
        _this.Id = id;
        _this.Description = description;
        _this.ExpandGroups = expandGroups;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    AssignmentStrategyDef.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Id = null;
    };
    /**
     * @return {?}
     */
    AssignmentStrategyDef.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new AssignmentStrategyDef(this.Model);
    };
    /**
     * @return {?}
     */
    AssignmentStrategyDef.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "Id", this.Id);
        BpmnModelHelper.SetAttributeValue(obj, "Description", this.Description);
        BpmnModelHelper.SetAttributeValue(obj, "ExpandGroups", this.ExpandGroups);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AssignmentStrategyDef.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Id = obj.Id;
        this.Description = obj.Description;
        this.ExpandGroups = obj.ExpandGroups;
    };
    return AssignmentStrategyDef;
}(ExtensionElement));
export { AssignmentStrategyDef };
if (false) {
    /** @type {?} */
    AssignmentStrategyDef.prototype.ClrTypeID;
    /** @type {?} */
    AssignmentStrategyDef.prototype.Id;
    /** @type {?} */
    AssignmentStrategyDef.prototype.Description;
    /** @type {?} */
    AssignmentStrategyDef.prototype.ExpandGroups;
}
BpmnModelHelper.RegisterElementType(new AssignmentStrategyDef(new BpmnModel()));
