// @farris/ide-devkit/esm5/lib/components/modules.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { BsDropdownModule } from './dropdown';
import { NgbTabsetModule } from './tabset/tabset.module';
import { NgbNavTabsetModule } from './navtab/navtab.module';
import { NgbAccordionModule } from './accordion/accordion.module';
import { PanelModule } from './panel/panel.module';
/** @type {?} */
var MODULES = [
    BsDropdownModule
];
/** @type {?} */
var NGB_MODULES = [
    NgbTabsetModule,
    NgbNavTabsetModule,
    NgbAccordionModule
];
/** @type {?} */
var IDE_MODULES = [
    PanelModule
];
// @NgModule({
//   imports: [
//     BsDropdownModule.forRoot()
//   ],
//   exports: [BsDropdownModule],
//   schemas: [NO_ERRORS_SCHEMA]
// })
// export class MDBRootModule {
// }
// @NgModule({ exports: [BsDropdownModule] })
// export class MDBBootstrapModule {
//   public static forRoot(): ModuleWithProviders {
//     return { ngModule: MDBRootModule };
//   }
// }
var NgbRootModule = /** @class */ (function () {
    function NgbRootModule() {
    }
    NgbRootModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        NgbTabsetModule.forRoot(),
                        NgbNavTabsetModule.forRoot(),
                        NgbAccordionModule.forRoot()
                    ],
                    exports: [
                        NGB_MODULES
                    ]
                },] }
    ];
    return NgbRootModule;
}());
export { NgbRootModule };
var NgbModule = /** @class */ (function () {
    function NgbModule() {
    }
    /**
     * @return {?}
     */
    NgbModule.forRoot = /**
     * @return {?}
     */
    function () { return { ngModule: NgbRootModule }; };
    NgbModule.decorators = [
        { type: NgModule, args: [{},] }
    ];
    return NgbModule;
}());
export { NgbModule };
var IDERootModule = /** @class */ (function () {
    function IDERootModule() {
    }
    IDERootModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        PanelModule.forRoot()
                    ]
                },] }
    ];
    return IDERootModule;
}());
export { IDERootModule };
