// @progress/kendo-angular-layout/dist/es2015/tabstrip/tabstrip-tab.component.js
import { Component, ContentChild, Input, ContentChildren, QueryList } from '@angular/core';
import { TabContentDirective } from './tab-content.directive';
import { TabTitleDirective } from './tab-title.directive';
/**
 * Represents the tab component of the TabStrip.
 */
export class TabStripTabComponent {
    constructor() {
        this.active = false;
        this._tabContent = new QueryList();
    }
    get tabContent() {
        return this._tabContent.first;
    }
    ngAfterContentInit() {
        this.active = this.selected;
    }
    ngOnChanges(changes) {
        if (changes['selected'] && !changes['selected'].isFirstChange()) {
            this.active = this.selected;
        }
    }
}
TabStripTabComponent.decorators = [
    { type: Component, args: [{
                exportAs: 'kendoTabStripTab',
                selector: 'kendo-tabstrip-tab',
                template: ``
            },] },
];
/** @nocollapse */
TabStripTabComponent.ctorParameters = () => [];
TabStripTabComponent.propDecorators = {
    'title': [{ type: Input },],
    'disabled': [{ type: Input },],
    'cssClass': [{ type: Input },],
    'selected': [{ type: Input },],
    '_tabContent': [{ type: ContentChildren, args: [TabContentDirective,] },],
    'tabTitle': [{ type: ContentChild, args: [TabTitleDirective,] },],
};
