// @farris/devkit/esm2015/lib/expression/expression_result_factory.js
import { Injectable, Injector } from "@angular/core";
import { ExpressionManager } from "./expression_manager";
;
import { ExpressionRegistry } from "./expression_registry";
import { ExpressionResult } from "./expression_result";
export class ExpressionResultFactory {
    constructor(injector, expressionRegistry, expressionManager, expressionResult) {
        this.injector = injector;
        this.expressionRegistry = expressionRegistry;
        this.expressionManager = expressionManager;
        this.expressionResult = expressionResult;
        this.registeEvent();
    }
    registeEvent() {
        this.expressionRegistry.expressions.subscribe((expressionObjects) => {
            // 加载完表达式之后做一次计算
            expressionObjects.forEach((expressionObject) => {
                if (expressionObject.deps && expressionObject.deps.length > 0) {
                    return;
                }
                const result = this.expressionManager.eval(expressionObject.id);
                this.expressionResult[expressionObject.id] = result;
            });
        });
    }
}
ExpressionResultFactory.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ExpressionResultFactory.ctorParameters = () => [
    { type: Injector },
    { type: ExpressionRegistry },
    { type: ExpressionManager },
    { type: ExpressionResult }
];
