// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/template-assignment-serializer/dimension-info-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/template-assignment-serializer/dimension-info-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DimensionInfo } from '../../param-template-assignment-vo/dimension-info';
import { DimensionSerializer } from '../dimension-serializer/dimension-serializer';
import { DimensionDataSerializer } from './dimension-data-serializer';
var DimensionInfoSerializer = /** @class */ (function () {
    function DimensionInfoSerializer() {
    }
    /**
     * @param {?} dimensionInfo
     * @return {?}
     */
    DimensionInfoSerializer.prototype.serialize = /**
     * @param {?} dimensionInfo
     * @return {?}
     */
    function (dimensionInfo) {
        if (!dimensionInfo) {
            return null;
        }
        /** @type {?} */
        var jobj = new Object();
        jobj['dataModificationAllow'] = dimensionInfo.dataModificationAllow;
        if (dimensionInfo.data) {
            jobj['data'] = dimensionInfo.data.map((/**
             * @param {?} dimensionData
             * @return {?}
             */
            function (dimensionData) { return new DimensionDataSerializer().serialize(dimensionData); }));
        }
        if (dimensionInfo.definition) {
            /** @type {?} */
            var serializer = new DimensionSerializer();
            jobj['definition'] = serializer.serialize(dimensionInfo.definition);
        }
        return jobj;
    };
    /**
     * @param {?} jobj
     * @return {?}
     */
    DimensionInfoSerializer.prototype.deserialize = /**
     * @param {?} jobj
     * @return {?}
     */
    function (jobj) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        var dimensionInfo = new DimensionInfo();
        dimensionInfo.dataModificationAllow = (/** @type {?} */ (jobj['dataModificationAllow']));
        if (jobj['data']) {
            dimensionInfo.data = jobj['data'].map((/**
             * @param {?} dimensionData
             * @return {?}
             */
            function (dimensionData) { return new DimensionDataSerializer().deserialize(dimensionData); }));
        }
        if (jobj['definition']) {
            /** @type {?} */
            var serializer = new DimensionSerializer();
            dimensionInfo.definition = serializer.deserialize(jobj['definition']);
        }
        return dimensionInfo;
    };
    return DimensionInfoSerializer;
}());
export { DimensionInfoSerializer };
