// @gsp-cmp/ubcc-param-vo/esm5/lib/vo/serializer/template-assignment-serializer/template-assignment-serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/vo/serializer/template-assignment-serializer/template-assignment-serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { UbccParamTemplateAssignment } from '../../param-template-assignment-vo/ubcc-param-template-assignment';
import { DimensionInfoSerializer } from './dimension-info-serializer';
import { EventInfo } from '../../param-template-assignment-vo/eventInfo';
var UbccParamTemplateAssignmentSerializer = /** @class */ (function () {
    function UbccParamTemplateAssignmentSerializer() {
    }
    /**
     * @param {?} templateAssignment
     * @return {?}
     */
    UbccParamTemplateAssignmentSerializer.prototype.serialize = /**
     * @param {?} templateAssignment
     * @return {?}
     */
    function (templateAssignment) {
        if (!templateAssignment) {
            return null;
        }
        /** @type {?} */
        var jobj = new Object();
        jobj['id'] = templateAssignment.id;
        jobj['templateId'] = templateAssignment.templateId;
        jobj['index'] = templateAssignment.index;
        jobj['templateCode'] = templateAssignment.templateCode;
        jobj['templateName'] = templateAssignment.templateName;
        jobj['templateType'] = templateAssignment.templateType;
        jobj['bizTypeId'] = templateAssignment.bizTypeId;
        jobj['useEvent'] = templateAssignment.useEvent;
        jobj['eventInfo'] = templateAssignment.eventInfo;
        jobj['useDimension'] = templateAssignment.useDimension;
        if (templateAssignment.dimensionInfo) {
            /** @type {?} */
            var serializer = new DimensionInfoSerializer();
            jobj['dimensionInfo'] = serializer.serialize(templateAssignment.dimensionInfo);
        }
        return jobj;
    };
    /**
     * @param {?} jobj
     * @return {?}
     */
    UbccParamTemplateAssignmentSerializer.prototype.deserialize = /**
     * @param {?} jobj
     * @return {?}
     */
    function (jobj) {
        if (!jobj) {
            return null;
        }
        /** @type {?} */
        var templateAssignment = new UbccParamTemplateAssignment();
        templateAssignment.id = jobj['id'];
        templateAssignment.templateId = jobj['templateId'];
        templateAssignment.index = (/** @type {?} */ (jobj['index']));
        templateAssignment.templateCode = jobj['templateCode'];
        templateAssignment.templateName = jobj['templateName'];
        templateAssignment.templateType = jobj['templateType'];
        templateAssignment.bizTypeId = jobj['bizTypeId'];
        templateAssignment.useEvent = (/** @type {?} */ (jobj['useEvent']));
        templateAssignment.eventInfo = Object.assign(new EventInfo(), jobj['eventInfo']);
        templateAssignment.useDimension = (/** @type {?} */ (jobj['useDimension']));
        if (jobj['dimensionInfo']) {
            /** @type {?} */
            var serializer = new DimensionInfoSerializer();
            templateAssignment.dimensionInfo = serializer.deserialize(jobj['dimensionInfo']);
        }
        return templateAssignment;
    };
    return UbccParamTemplateAssignmentSerializer;
}());
export { UbccParamTemplateAssignmentSerializer };
