// @gspwidget/util/esm2015/lib/types/data.type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { camelCase, constantCase } from 'change-case';
import { decrypt, encrypt } from '../utils/encryption';
/**
 * @record
 */
export function ProviderParam() { }
if (false) {
    /** @type {?|undefined} */
    ProviderParam.prototype.label;
    /** @type {?|undefined} */
    ProviderParam.prototype.name;
    /** @type {?|undefined} */
    ProviderParam.prototype.type;
    /** @type {?|undefined} */
    ProviderParam.prototype.sortNo;
    /** @type {?|undefined} */
    ProviderParam.prototype.value;
    /** @type {?|undefined} */
    ProviderParam.prototype.placeholder;
    /** @type {?|undefined} */
    ProviderParam.prototype.options;
    /** @type {?|undefined} */
    ProviderParam.prototype.checked;
    /** @type {?|undefined} */
    ProviderParam.prototype.required;
    /** @type {?|undefined} */
    ProviderParam.prototype.encrypted;
}
// TODO: 数据源相关的，要放在 dashboard core 中吗
export class Datasrc {
    /**
     * @param {?=} providerName
     */
    constructor(providerName) {
        this.providerName = providerName;
        this.i18nName = {};
        this.config = {};
    }
}
if (false) {
    /** @type {?} */
    Datasrc.prototype.id;
    /** @type {?} */
    Datasrc.prototype.name;
    /** @type {?} */
    Datasrc.prototype.i18nName;
    /** @type {?} */
    Datasrc.prototype.config;
    /** @type {?} */
    Datasrc.prototype.groupName;
    /** @type {?} */
    Datasrc.prototype.creator;
    /** @type {?} */
    Datasrc.prototype.createTime;
    /** @type {?} */
    Datasrc.prototype.providerName;
}
/**
 * @record
 */
export function DataProvider() { }
if (false) {
    /** @type {?} */
    DataProvider.prototype.label;
    /** @type {?} */
    DataProvider.prototype.name;
}
/**
 * @record
 */
export function DataParamDto() { }
if (false) {
    /** @type {?|undefined} */
    DataParamDto.prototype.id;
    /** @type {?|undefined} */
    DataParamDto.prototype.i18nName;
    /** @type {?|undefined} */
    DataParamDto.prototype.type;
    /** @type {?|undefined} */
    DataParamDto.prototype.config;
    /** @type {?|undefined} */
    DataParamDto.prototype.datasetId;
    /** @type {?|undefined} */
    DataParamDto.prototype.name;
    /** @type {?|undefined} */
    DataParamDto.prototype.value;
}
/**
 * 取数参数
 */
export class DataParam {
    constructor() {
        this.type = 'QUERY_PARAM';
        this.default = null;
    }
    /**
     * @param {?} dp
     * @return {?}
     */
    static ToDto(dp) {
        const { i18nName, type, config, id } = dp;
        return (/** @type {?} */ ({
            id, i18nName, config,
            type: camelCase(type),
            value: dp.default,
        }));
    }
    /**
     * @param {?} dpDto
     * @return {?}
     */
    static FromDto(dpDto) {
        return (/** @type {?} */ ({
            id: dpDto.id,
            i18nName: dpDto.i18nName,
            type: constantCase(dpDto.type),
            config: dpDto.config,
            datasetId: dpDto.datasetId,
            name: dpDto.name,
            default: dpDto.value,
        }));
    }
}
if (false) {
    /** @type {?} */
    DataParam.prototype.id;
    /** @type {?} */
    DataParam.prototype.name;
    /** @type {?} */
    DataParam.prototype.i18nName;
    /** @type {?} */
    DataParam.prototype.type;
    /** @type {?} */
    DataParam.prototype.config;
    /** @type {?} */
    DataParam.prototype.default;
    /** @type {?} */
    DataParam.prototype.datasetId;
}
/**
 * @record
 */
export function DatasetDto() { }
if (false) {
    /** @type {?|undefined} */
    DatasetDto.prototype.id;
    /** @type {?|undefined} */
    DatasetDto.prototype.name;
    /** @type {?|undefined} */
    DatasetDto.prototype.datasourceId;
    /** @type {?|undefined} */
    DatasetDto.prototype.i18nName;
    /** @type {?|undefined} */
    DatasetDto.prototype.datasourceName;
    /** @type {?|undefined} */
    DatasetDto.prototype.dimensions;
    /** @type {?|undefined} */
    DatasetDto.prototype.creationTime;
    /** @type {?|undefined} */
    DatasetDto.prototype.method;
    /** @type {?|undefined} */
    DatasetDto.prototype.parameters;
    /** @type {?|undefined} */
    DatasetDto.prototype.config;
    /** @type {?|undefined} */
    DatasetDto.prototype.dataParamDto;
}
export class DatasetParameter {
    /**
     * @param {?} name
     * @param {?} value
     * @param {?=} sortNo
     * @param {?=} isExpression
     */
    constructor(name, value, sortNo, isExpression = true) {
        this.name = name;
        this.value = value;
        this.sortNo = sortNo;
        this.isExpression = isExpression;
    }
}
if (false) {
    /** @type {?} */
    DatasetParameter.prototype.name;
    /** @type {?} */
    DatasetParameter.prototype.value;
    /** @type {?} */
    DatasetParameter.prototype.sortNo;
    /** @type {?} */
    DatasetParameter.prototype.isExpression;
}
/**
 * @record
 */
function AnyKey() { }
export class Dataset {
    constructor() {
        // datasrc?: Datasrc
        this.config = {};
        this.cols = {
            dimensions: [],
            measures: [],
        };
        this.dataParams = [];
    }
    /**
     * @return {?}
     */
    static create() {
        return new Dataset();
    }
    /**
     * @param {?} dataset
     * @return {?}
     */
    static toDto(dataset) {
        const { id, name, datasrcId, datasrcName, cols, creationTime, config, dataParams, i18nName } = dataset;
        return {
            id, name, creationTime, i18nName,
            datasourceId: datasrcId,
            datasourceName: datasrcName,
            dimensions: cols.dimensions,
            config: encrypt(JSON.stringify(config)),
            dataParamDto: dataParams.map(DataParam.ToDto)
        };
    }
    /**
     * @param {?} datasetDto
     * @return {?}
     */
    static fromDto(datasetDto) {
        const { id, name, datasourceId, datasourceName, dimensions, creationTime, method, parameters, i18nName, config, dataParamDto } = datasetDto;
        /** @type {?} */
        let parsedConfig = null;
        try {
            parsedConfig = JSON.parse(decrypt(config));
        }
        catch (error) {
        }
        if (!parsedConfig)
            parsedConfig = {};
        return {
            // ...datasetDto,
            id, name, creationTime, i18nName,
            datasrcId: datasourceId,
            datasrcName: datasourceName,
            cols: { dimensions },
            config: parsedConfig,
            dataParams: dataParamDto ? dataParamDto.map(DataParam.FromDto) : [],
        };
    }
}
if (false) {
    /** @type {?} */
    Dataset.prototype.id;
    /** @type {?} */
    Dataset.prototype.name;
    /** @type {?} */
    Dataset.prototype.product;
    /** @type {?} */
    Dataset.prototype.module;
    /** @type {?} */
    Dataset.prototype.creator;
    /** @type {?} */
    Dataset.prototype.modifier;
    /** @type {?} */
    Dataset.prototype.creationTime;
    /** @type {?} */
    Dataset.prototype.modifyTime;
    /** @type {?} */
    Dataset.prototype.datasrcId;
    /** @type {?} */
    Dataset.prototype.datasrcName;
    /** @type {?} */
    Dataset.prototype.i18nName;
    /** @type {?} */
    Dataset.prototype.config;
    /** @type {?} */
    Dataset.prototype.cols;
    /** @type {?} */
    Dataset.prototype.dataParams;
}
export class Dimension {
}
if (false) {
    /** @type {?} */
    Dimension.prototype.name;
    /** @type {?} */
    Dimension.prototype.label;
    /** @type {?} */
    Dimension.prototype.i18nLabel;
    /** @type {?} */
    Dimension.prototype.type;
    /** @type {?} */
    Dimension.prototype.typeName;
    /** @type {?} */
    Dimension.prototype.javaClassName;
    /** @type {?} */
    Dimension.prototype.groupName;
    /** @type {?} */
    Dimension.prototype.comment;
    /** @type {?} */
    Dimension.prototype.notNull;
}
/**
 * @record
 */
export function Filter() { }
if (false) {
    /** @type {?|undefined} */
    Filter.prototype.hasValue;
    /** @type {?|undefined} */
    Filter.prototype.filterExpression;
}
/** @enum {string} */
const ComputedType = {
    Count: 'Count',
    Sum: 'Sum',
    Max: 'Max',
    Min: 'Min',
    Avg: 'Avg',
    Expression: 'Expression',
};
export { ComputedType };
/** @enum {string} */
const AggType = {
    Table: 'Table',
    CroosTable: 'CroosTable',
};
export { AggType };
export class ComputedCol extends Dimension {
    /**
     * @param {?=} expression
     */
    constructor(expression) {
        super();
        this.expression = expression;
        this.computeType = ComputedType.Expression;
    }
}
if (false) {
    /** @type {?} */
    ComputedCol.prototype.computeType;
    /**
     * @type {?}
     * @private
     */
    ComputedCol.prototype.expression;
}
/**
 * @record
 */
export function Col() { }
/**
 * @record
 */
export function Row() { }
/**
 * @record
 */
export function AggDto() { }
if (false) {
    /** @type {?|undefined} */
    AggDto.prototype.id;
    /** @type {?|undefined} */
    AggDto.prototype.dataSetId;
    /** @type {?|undefined} */
    AggDto.prototype.dataSetName;
    /** @type {?|undefined} */
    AggDto.prototype.resource;
    /** @type {?|undefined} */
    AggDto.prototype.resourceId;
    /** @type {?|undefined} */
    AggDto.prototype.aggregationType;
    /** @type {?|undefined} */
    AggDto.prototype.columnConfigs;
    /** @type {?|undefined} */
    AggDto.prototype.rowConfigs;
    /** @type {?|undefined} */
    AggDto.prototype.filterConfigs;
    /** @type {?|undefined} */
    AggDto.prototype.computedExpressionConfigs;
}
export class Agg {
    constructor() {
        this.cols = [];
        this.rows = [];
        this.filters = [];
        this.computedCols = [];
    }
    /**
     * @param {?} agg
     * @return {?}
     */
    static toDto(agg) {
        const { id, datasetId, resource, resourceId, aggType, cols, rows, filters, computedCols } = agg;
        return (/** @type {?} */ ({
            id,
            dataSetId: datasetId,
            resource,
            resourceId,
            aggregationType: aggType,
            columnConfigs: cols,
            rowConfigs: rows,
            filterConfigs: filters,
            computedExpressionConfigs: computedCols
        }));
    }
    /**
     * @param {?} aggDto
     * @return {?}
     */
    static fromDto(aggDto) {
        const { id, dataSetId, dataSetName, resource, resourceId, aggregationType, columnConfigs, rowConfigs, filterConfigs, computedExpressionConfigs } = aggDto;
        return (/** @type {?} */ ({
            id,
            datasetId: dataSetId,
            datasetName: dataSetName,
            resource,
            resourceId,
            aggType: aggregationType,
            cols: columnConfigs,
            rows: rowConfigs,
            filters: filterConfigs,
            computedCols: computedExpressionConfigs
        }));
    }
}
if (false) {
    /** @type {?} */
    Agg.prototype.id;
    /** @type {?} */
    Agg.prototype.datasetId;
    /** @type {?} */
    Agg.prototype.datasetName;
    /** @type {?} */
    Agg.prototype.resource;
    /** @type {?} */
    Agg.prototype.resourceId;
    /** @type {?} */
    Agg.prototype.aggType;
    /** @type {?} */
    Agg.prototype.cols;
    /** @type {?} */
    Agg.prototype.rows;
    /** @type {?} */
    Agg.prototype.filters;
    /** @type {?} */
    Agg.prototype.computedCols;
}
/** @enum {string} */
const AggregationResultSetType = {
    Table: 'Table',
    CroosTable: 'CroosTable',
};
export { AggregationResultSetType };
/**
 * 数据服务实际返回的结构
 * @record
 */
export function AggResult() { }
if (false) {
    /** @type {?|undefined} */
    AggResult.prototype.tableData;
    /** @type {?|undefined} */
    AggResult.prototype.dimensions;
    /** @type {?|undefined} */
    AggResult.prototype.columns;
    /**
     * ?
     * @type {?|undefined}
     */
    AggResult.prototype.type;
    /**
     * ?
     * @type {?|undefined}
     */
    AggResult.prototype.getResultSet;
    /**
     * ?
     * @type {?|undefined}
     */
    AggResult.prototype.data;
}
/**
 * @record
 */
export function Result() { }
if (false) {
    /** @type {?} */
    Result.prototype.data;
    /** @type {?} */
    Result.prototype.columns;
}
/**
 * widget component 需要外部传入的数据格式。上面的服务符合此结构。
 * @record
 */
export function DataWithDims() { }
if (false) {
    /** @type {?|undefined} */
    DataWithDims.prototype.dimensions;
    /** @type {?|undefined} */
    DataWithDims.prototype.tableData;
}
/**
 * 获取数据时传服务端的 dp + value
 */
export class DpValue {
    /**
     * @param {?} dpValue
     * @return {?}
     */
    static toDto(dpValue) {
        return (/** @type {?} */ ({
            dataParamId: dpValue.dpId,
            value: dpValue.value
        }));
    }
}
if (false) {
    /** @type {?} */
    DpValue.prototype.dpId;
    /** @type {?} */
    DpValue.prototype.value;
}
/**
 * @record
 */
export function DpValueDto() { }
if (false) {
    /** @type {?} */
    DpValueDto.prototype.dataParamId;
    /** @type {?} */
    DpValueDto.prototype.value;
}
/**
 * @record
 */
export function RuntimeAggOptionDto() { }
if (false) {
    /** @type {?} */
    RuntimeAggOptionDto.prototype.dimensions;
    /** @type {?} */
    RuntimeAggOptionDto.prototype.indicators;
}
export class Indicator {
    constructor() {
        this.aggFn = 4;
    }
}
if (false) {
    /** @type {?} */
    Indicator.prototype.dimName;
    /** @type {?} */
    Indicator.prototype.aggFn;
}
export class RuntimeAggOption {
    constructor() {
        this.dimensions = [];
        this.indicators = [new Indicator()];
    }
    /**
     * @param {?} runtimeAgg
     * @return {?}
     */
    static toDto(runtimeAgg) {
        if (!runtimeAgg)
            return;
        return (/** @type {?} */ ({
            dimensions: runtimeAgg.dimensions,
            indicators: runtimeAgg.indicators.map((/**
             * @param {?} ind
             * @return {?}
             */
            ind => ({
                type: ind.aggFn,
                name: ind.dimName
            })))
        }));
    }
}
if (false) {
    /** @type {?} */
    RuntimeAggOption.prototype.dimensions;
    /** @type {?} */
    RuntimeAggOption.prototype.indicators;
}
