// @gspwidget/util/esm2015/lib/types/igix.type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function AppOptions() { }
if (false) {
    /**
     * 指定当前打开的是菜单（Menu）还是应用（App）
     * @type {?}
     */
    AppOptions.prototype.appType;
    /**
     * 如果打开菜单，funcId需要指定，appId、appEntrance可以为空
     * @type {?}
     */
    AppOptions.prototype.funcId;
    /**
     * 如果打开应用，appId及appEntrance需要指定，funcId可以为空
     * @type {?}
     */
    AppOptions.prototype.appId;
    /** @type {?} */
    AppOptions.prototype.appEntrance;
    /** @type {?|undefined} */
    AppOptions.prototype.tabId;
    /** @type {?|undefined} */
    AppOptions.prototype.tabName;
    /** @type {?|undefined} */
    AppOptions.prototype.routeUrl;
    /** @type {?|undefined} */
    AppOptions.prototype.entityParams;
    /** @type {?|undefined} */
    AppOptions.prototype.queryStringParams;
    /** @type {?|undefined} */
    AppOptions.prototype.isReload;
    /** @type {?|undefined} */
    AppOptions.prototype.isJquery;
}
/**
 * @record
 */
export function BO() { }
if (false) {
    /** @type {?} */
    BO.prototype.id;
    /** @type {?} */
    BO.prototype.code;
    /** @type {?} */
    BO.prototype.name;
    /** @type {?} */
    BO.prototype.description;
    /** @type {?} */
    BO.prototype.sysInit;
    /** @type {?} */
    BO.prototype.sortOrder;
    /** @type {?} */
    BO.prototype.isDetail;
    /** @type {?} */
    BO.prototype.layer;
    /** @type {?} */
    BO.prototype.type;
    /** @type {?} */
    BO.prototype.parentID;
}
/**
 * @record
 */
export function RoleGroup() { }
if (false) {
    /** @type {?} */
    RoleGroup.prototype.id;
    /** @type {?} */
    RoleGroup.prototype.code;
    /** @type {?} */
    RoleGroup.prototype.name;
    /** @type {?} */
    RoleGroup.prototype.path;
    /** @type {?} */
    RoleGroup.prototype.layer;
    /** @type {?} */
    RoleGroup.prototype.isDetail;
    /** @type {?} */
    RoleGroup.prototype.sysInit;
    /** @type {?} */
    RoleGroup.prototype.note;
    /** @type {?} */
    RoleGroup.prototype.parentId;
    /** @type {?} */
    RoleGroup.prototype.sortOrder;
    /** @type {?} */
    RoleGroup.prototype.creator;
    /** @type {?} */
    RoleGroup.prototype.createdTime;
    /** @type {?} */
    RoleGroup.prototype.lastModifier;
    /** @type {?} */
    RoleGroup.prototype.lastModifiedTime;
    /** @type {?} */
    RoleGroup.prototype.farris_selectable;
}
/**
 * @record
 */
export function Role() { }
if (false) {
    /** @type {?} */
    Role.prototype.id;
    /** @type {?} */
    Role.prototype.code;
    /** @type {?} */
    Role.prototype.name;
    /** @type {?} */
    Role.prototype.roleGroup;
    /** @type {?} */
    Role.prototype.roleGroupCode;
    /** @type {?} */
    Role.prototype.roleGroupName;
    /** @type {?} */
    Role.prototype.note;
    /** @type {?} */
    Role.prototype.sysInit;
    /** @type {?} */
    Role.prototype.state;
    /** @type {?} */
    Role.prototype.validFrom;
    /** @type {?} */
    Role.prototype.validTo;
    /** @type {?} */
    Role.prototype.createdTime;
    /** @type {?} */
    Role.prototype.creator;
    /** @type {?} */
    Role.prototype.creatorCode;
    /** @type {?} */
    Role.prototype.creatorName;
    /** @type {?} */
    Role.prototype.lastModifiedTime;
    /** @type {?} */
    Role.prototype.lastModifier;
    /** @type {?} */
    Role.prototype.lastModifierCode;
    /** @type {?} */
    Role.prototype.lastModifierName;
    /** @type {?|undefined} */
    Role.prototype.isRole;
}
/**
 * @record
 */
export function Operation() { }
if (false) {
    /** @type {?|undefined} */
    Operation.prototype.id;
    /** @type {?|undefined} */
    Operation.prototype.name;
}
/**
 * 来自家奇的 /api/runtime/sys/v1.0/userinfos?infoType=user 接口
 * @record
 */
export function UserInfo() { }
if (false) {
    /** @type {?} */
    UserInfo.prototype.address;
    /** @type {?} */
    UserInfo.prototype.code;
    /** @type {?} */
    UserInfo.prototype.email;
    /** @type {?} */
    UserInfo.prototype.id;
    /** @type {?} */
    UserInfo.prototype.idCard;
    /** @type {?} */
    UserInfo.prototype.lanName;
    /** @type {?} */
    UserInfo.prototype.languageId;
    /** @type {?} */
    UserInfo.prototype.mobilePhone;
    /** @type {?} */
    UserInfo.prototype.name;
    /** @type {?} */
    UserInfo.prototype.orgId;
    /** @type {?} */
    UserInfo.prototype.orgName;
    /** @type {?} */
    UserInfo.prototype.phone;
    /** @type {?} */
    UserInfo.prototype.qq;
    /** @type {?} */
    UserInfo.prototype.sex;
    /** @type {?} */
    UserInfo.prototype.tenantId;
    /** @type {?} */
    UserInfo.prototype.tenantName;
    /** @type {?} */
    UserInfo.prototype.userId;
    /** @type {?} */
    UserInfo.prototype.userSetting;
    /** @type {?} */
    UserInfo.prototype.weChat;
}
/**
 * @record
 */
export function UserSetting() { }
if (false) {
    /** @type {?} */
    UserSetting.prototype.id;
    /** @type {?} */
    UserSetting.prototype.img;
    /** @type {?} */
    UserSetting.prototype.imgblob;
    /** @type {?} */
    UserSetting.prototype.userId;
}
/**
 * @record
 */
export function InvokeInfo() { }
if (false) {
    /** @type {?} */
    InvokeInfo.prototype.Type;
    /** @type {?} */
    InvokeInfo.prototype.Path;
    /** @type {?} */
    InvokeInfo.prototype.Url;
    /** @type {?} */
    InvokeInfo.prototype.Image;
    /** @type {?} */
    InvokeInfo.prototype.Width;
    /** @type {?} */
    InvokeInfo.prototype.Height;
}
/**
 * @record
 */
export function RtfFunc() { }
if (false) {
    /** @type {?} */
    RtfFunc.prototype.id;
    /** @type {?} */
    RtfFunc.prototype.code;
    /** @type {?} */
    RtfFunc.prototype.name;
    /** @type {?} */
    RtfFunc.prototype.tenantId;
    /** @type {?} */
    RtfFunc.prototype.isModule;
    /** @type {?} */
    RtfFunc.prototype.isDetail;
    /** @type {?} */
    RtfFunc.prototype.isSysMenu;
    /** @type {?} */
    RtfFunc.prototype.parentId;
    /** @type {?} */
    RtfFunc.prototype.sortOrder;
    /** @type {?} */
    RtfFunc.prototype.bizObjectId;
    /** @type {?} */
    RtfFunc.prototype.bizOpId;
    /** @type {?} */
    RtfFunc.prototype.icon;
    /** @type {?} */
    RtfFunc.prototype.invokingConfig;
    /** @type {?} */
    RtfFunc.prototype.isWorkStation;
    /** @type {?} */
    RtfFunc.prototype.lastModifiedTime;
    /** @type {?} */
    RtfFunc.prototype.lastModifier;
    /** @type {?} */
    RtfFunc.prototype.creator;
    /** @type {?} */
    RtfFunc.prototype.createdTime;
    /** @type {?} */
    RtfFunc.prototype.funcType;
    /** @type {?} */
    RtfFunc.prototype.path;
    /** @type {?} */
    RtfFunc.prototype.layer;
    /** @type {?} */
    RtfFunc.prototype.description;
    /** @type {?} */
    RtfFunc.prototype.child;
}
/** @enum {string} */
const FuncType = {
    Products: '1',
    Module: '2',
    AppGroups: '3',
    App: '4',
};
export { FuncType };
