// @gspwidget/util/esm2015/lib/types/portlet.type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const DEFAULT_PORTLET_COLS = 5;
/** @type {?} */
export const DEFAULT_PORTLET_ROWS = 4;
/**
 * 部件
 * @abstract
 */
export class Portlet {
    /**
     * 唯一标识
     * @param {?} key
     * @return {?}
     */
    getSize(key) {
        if (key === 'rows') {
            return this._getAxisSize('rows') || this._getAxisSize('minRows') || DEFAULT_PORTLET_ROWS;
        }
        else if (key === 'cols') {
            return this._getAxisSize('cols') || this._getAxisSize('minCols') || DEFAULT_PORTLET_COLS;
        }
        else {
            return this._getAxisSize(key);
        }
    }
}
if (false) {
    /** @type {?} */
    Portlet.prototype.id;
    /**
     * @abstract
     * @param {?} key
     * @return {?}
     */
    Portlet.prototype._getAxisSize = function (key) { };
}
export class EppPortlet extends Portlet {
    /**
     * @param {?} widgetId
     */
    constructor(widgetId) {
        super();
        this.widgetId = widgetId;
        this.name = '未命名部件'; /** 显示标题 */
        /**
         * 显示标题
         */
        this.i18nName = {};
        this.propValues = {}; /** 可配置属性的值 */
        /**
         * 可配置属性的值
         */
        this.dataOption = {};
        /** 所属模块 */
        /**
         * 配置联动时，前端动态增加。只做运行时使用，不应该被保存。
         * null 代表初始状态，[] 代表数据已加载且为空。
         */
        this.dps = null;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    _getAxisSize(key) {
        return (this.size && this.size[key]) || (this.widget.layout && this.widget.layout.grid && this.widget.layout.grid[key]);
    }
}
if (false) {
    /** @type {?} */
    EppPortlet.prototype.name;
    /**
     * 显示标题
     * @type {?}
     */
    EppPortlet.prototype.i18nName;
    /** @type {?} */
    EppPortlet.prototype.desc;
    /**
     * 备注
     * @type {?}
     */
    EppPortlet.prototype.size;
    /**
     * 尺寸范围
     * @type {?}
     */
    EppPortlet.prototype.widget;
    /** @type {?} */
    EppPortlet.prototype.propValues;
    /**
     * 可配置属性的值
     * @type {?}
     */
    EppPortlet.prototype.dataOption;
    /** @type {?} */
    EppPortlet.prototype.moduleId;
    /**
     * 配置联动时，前端动态增加。只做运行时使用，不应该被保存。
     * null 代表初始状态，[] 代表数据已加载且为空。
     * @type {?}
     */
    EppPortlet.prototype.dps;
    /**
     * 元部件 Id
     * @type {?}
     */
    EppPortlet.prototype.widgetId;
}
/** @type {?} */
const defaultSizeMap = {
    rows: 4,
    cols: 5,
    minCols: 2,
    minRows: 2,
    maxCols: 24,
    maxRows: 24,
};
export class BAPortlet extends Portlet {
    constructor() {
        super(...arguments);
        this.isBA = true;
    }
    // code?: string
    // createTime?: string
    // detail?: number
    // gradeCode?: string
    // isParent?: boolean
    // layer?: number
    // pid?: string
    // tab?: number
    // type?: string
    // updateTime?: string
    // userId?: string
    /**
     * @param {?} key
     * @return {?}
     */
    _getAxisSize(key) {
        return defaultSizeMap[key];
    }
}
if (false) {
    /** @type {?} */
    BAPortlet.prototype.isBA;
    /** @type {?} */
    BAPortlet.prototype.name;
}
