// @gspwidget/util/esm2015/lib/types/widget.type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function BasicWidgetInfoAotMode() { }
if (false) {
    /** @type {?|undefined} */
    BasicWidgetInfoAotMode.prototype.name;
    /** @type {?|undefined} */
    BasicWidgetInfoAotMode.prototype.href;
    /** @type {?|undefined} */
    BasicWidgetInfoAotMode.prototype.baseUrl;
    /** @type {?} */
    BasicWidgetInfoAotMode.prototype.version;
}
/**
 * @record
 */
export function BasicWidgetInfoDevMode() { }
if (false) {
    /** @type {?} */
    BasicWidgetInfoDevMode.prototype.widgetComponent;
    /** @type {?|undefined} */
    BasicWidgetInfoDevMode.prototype.baseUrl;
    /** @type {?|undefined} */
    BasicWidgetInfoDevMode.prototype.name;
    /** @type {?} */
    BasicWidgetInfoDevMode.prototype.version;
}
/** @enum {number} */
const WidgetContainerMode = {
    AOT: 0,
    DEV: 1,
    JIT: 2,
};
export { WidgetContainerMode };
WidgetContainerMode[WidgetContainerMode.AOT] = 'AOT';
WidgetContainerMode[WidgetContainerMode.DEV] = 'DEV';
WidgetContainerMode[WidgetContainerMode.JIT] = 'JIT';
/**
 * @record
 */
export function Size() { }
if (false) {
    /** @type {?|undefined} */
    Size.prototype.minCols;
    /** @type {?|undefined} */
    Size.prototype.maxCols;
    /** @type {?|undefined} */
    Size.prototype.minRows;
    /** @type {?|undefined} */
    Size.prototype.maxRows;
}
/**
 * @record
 */
export function WidgetLayoutGridConfig() { }
if (false) {
    /** @type {?|undefined} */
    WidgetLayoutGridConfig.prototype.minCols;
    /** @type {?|undefined} */
    WidgetLayoutGridConfig.prototype.maxCols;
    /** @type {?|undefined} */
    WidgetLayoutGridConfig.prototype.minRows;
    /** @type {?|undefined} */
    WidgetLayoutGridConfig.prototype.maxRows;
    /** @type {?|undefined} */
    WidgetLayoutGridConfig.prototype.cols;
    /** @type {?|undefined} */
    WidgetLayoutGridConfig.prototype.rows;
}
/**
 * @record
 */
export function WidgetLayout() { }
if (false) {
    /** @type {?} */
    WidgetLayout.prototype.grid;
}
/**
 * @record
 */
export function WidgetInfo() { }
if (false) {
    /**
     * 注册后生成
     * @type {?}
     */
    WidgetInfo.prototype.id;
    /** @type {?} */
    WidgetInfo.prototype.widgetCode;
    /** @type {?} */
    WidgetInfo.prototype.href;
    /**
     * 来自元部件配置信息
     * @type {?}
     */
    WidgetInfo.prototype.su;
    /** @type {?} */
    WidgetInfo.prototype.i18nName;
    /** @type {?} */
    WidgetInfo.prototype.i18nDesc;
    /** @type {?} */
    WidgetInfo.prototype.name;
    /** @type {?|undefined} */
    WidgetInfo.prototype.version;
    /** @type {?} */
    WidgetInfo.prototype.displayName;
    /** @type {?|undefined} */
    WidgetInfo.prototype.desc;
    /** @type {?|undefined} */
    WidgetInfo.prototype.layout;
    /** @type {?|undefined} */
    WidgetInfo.prototype.logoImgUrl;
    /** @type {?|undefined} */
    WidgetInfo.prototype.previewImgUrl;
    /**
     * 元部件存放的根路径，注意这个结尾，没有 正斜杠 '/' ！
     * @type {?|undefined}
     */
    WidgetInfo.prototype.baseUrl;
}
/**
 * @record
 */
export function FilterWidgetEventParam() { }
if (false) {
    /** @type {?} */
    FilterWidgetEventParam.prototype.eventKey;
    /** @type {?} */
    FilterWidgetEventParam.prototype.value;
}
/**
 * @record
 */
export function WidgetEventConfig() { }
if (false) {
    /**
     * 可配置为任意显示名称，或国际化 code
     * @type {?|undefined}
     */
    WidgetEventConfig.prototype.name;
    /** @type {?|undefined} */
    WidgetEventConfig.prototype.desc;
    /**
     * 事件对应属性名，程序自动赋值，不需要配置
     * @type {?|undefined}
     */
    WidgetEventConfig.prototype.propName;
}
