// @farris/approval-comments/esm5/utils/api.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/api.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
var ApiService = /** @class */ (function () {
    function ApiService(http) {
        this.http = http;
    }
    /**
     * @param {?} param
     * @return {?}
     */
    ApiService.prototype.demoapi = /**
     * @param {?} param
     * @return {?}
     */
    function (param) {
        /** @type {?} */
        var options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        return this.http.post(this.baseURL + '/path', param, options);
    };
    // 获取审批意见接口
    // 获取审批意见接口
    /**
     * @param {?} param
     * @return {?}
     */
    ApiService.prototype.getApprovalCommentApi = 
    // 获取审批意见接口
    /**
     * @param {?} param
     * @return {?}
     */
    function (param) {
        /** @type {?} */
        var options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        // this.baseURL = '/api/dev/nocode/v1.0/micro-apps/getworkitemlogs';
        this.baseURL = '/api/no/code/runtime/v1.0/getworkitemlogs';
        return this.http.post(this.baseURL, param, options);
    };
    // 获取审批节点接口
    // 获取审批节点接口
    /**
     * @param {?} microAppId
     * @return {?}
     */
    ApiService.prototype.getAuditorsByMicroAppId = 
    // 获取审批节点接口
    /**
     * @param {?} microAppId
     * @return {?}
     */
    function (microAppId) {
        /** @type {?} */
        var options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        this.baseURL = '/api/dev/nocode/v1.0/micro-apps/getauditorsbyappid/' + microAppId;
        return this.http.get(this.baseURL, options);
    };
    ApiService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ApiService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return ApiService;
}());
export { ApiService };
if (false) {
    /** @type {?} */
    ApiService.prototype.baseURL;
    /**
     * @type {?}
     * @private
     */
    ApiService.prototype.http;
}
