// @edp-bif/common-api/esm2015/lib/content/impl/AbstractRtBillProcessContent.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/content/impl/AbstractRtBillProcessContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '../../ref';
import { ConditionEntity } from "../../entity/impl/ConditionEntity";
import { MappingModeEntity } from '../../entity/impl/MappingModeEntity';
import { SaveModeEntity } from '../../entity/impl/SaveModeEntity';
/**
 * 单据流内容类。
 * @param TEntity 单据流实体。
 * @abstract
 * @template TEntity
 */
export class AbstractRtBillProcessContent extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.entity = null;
        this.eventRouterConditionsContent = new Array();
        this.mappingConditionsContent = new Array();
        this.mappingModeContent = null;
        this.saveModeContent = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "entity", this.entity);
        BpmnModelJsonHelper.SetAttributeValue(obj, "eventRouterConditionsContent", this.eventRouterConditionsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingConditionsContent", this.mappingConditionsContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingModeContent", this.mappingModeContent);
        BpmnModelJsonHelper.SetAttributeValue(obj, "saveModeContent", this.saveModeContent);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.entity = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "entity", this.getEntityClass());
        this.eventRouterConditionsContent = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "eventRouterConditionsContent", ConditionEntity);
        this.mappingConditionsContent = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingConditionsContent", ConditionEntity);
        this.mappingModeContent = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "mappingModeContent", MappingModeEntity);
        this.saveModeContent = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "saveModeContent", SaveModeEntity);
    }
}
if (false) {
    /** @type {?} */
    AbstractRtBillProcessContent.prototype.entity;
    /** @type {?} */
    AbstractRtBillProcessContent.prototype.eventRouterConditionsContent;
    /** @type {?} */
    AbstractRtBillProcessContent.prototype.mappingConditionsContent;
    /** @type {?} */
    AbstractRtBillProcessContent.prototype.mappingModeContent;
    /** @type {?} */
    AbstractRtBillProcessContent.prototype.saveModeContent;
    /**
     * 获取泛型类类型。
     *
     * @abstract
     * @protected
     * @template ClassOfT
     * @return {?} 泛型类类型。
     */
    AbstractRtBillProcessContent.prototype.getEntityClass = function () { };
}
