// @gsp-bef/gsp-vo-metadata/fesm2015/gsp-bef-gsp-vo-metadata.js
import { GSPCommonModel, GSPCommonElement, GSPCommonObject } from '@gsp-bef/gsp-cm-metadata';
export { GSPCommonElement as ɵb, GSPCommonModel as ɵa, GSPCommonObject as ɵc } from '@gsp-bef/gsp-cm-metadata';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 视图对象类
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 视图对象类
 */
class GSPViewModel extends GSPCommonModel {
    /**
     * 获取扩展类型,总是返回为'视图对象'类型
     * @return {?}
     */
    get ExtendType() {
        return 'GSPViewModel';
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * VO节点元素
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * VO节点元素
 */
class GSPViewModelElement extends GSPCommonElement {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * BE节点实体类
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * BE节点实体类
 */
class GSPViewObject extends GSPCommonObject {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 数据逻辑扩展
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 数据逻辑扩展
 */
class VoDataExtendInfo {
}
/** @type {?} */
const VoDataExtendInfos = [{
        Name: '扩展操作',
        Value: 'DataExtendInfo',
        items: [
            {
                Index: 0,
                Name: '数据Mapping',
                Value: 'DataMappingActions',
                HasInternalAction: true
            },
            {
                Index: 1,
                Name: '查询数据前',
                Value: 'BeforeQueryActions',
                HasInternalAction: false
            },
            {
                Index: 2,
                Name: '查询数据',
                Value: 'QueryActions',
                HasInternalAction: true
            },
            {
                Index: 3,
                Name: '查询数据后',
                Value: 'AfterQueryActions',
                HasInternalAction: false
            },
            {
                Index: 4,
                Name: '检索数据前',
                Value: 'BeforeRetrieveActions',
                HasInternalAction: false
            },
            {
                Index: 5,
                Name: '检索数据',
                Value: 'RetrieveActions',
                HasInternalAction: true
            },
            {
                Index: 6,
                Name: '检索数据后',
                Value: 'AfterRetrieveActions',
                HasInternalAction: false
            },
            {
                Index: 7,
                Name: '修改数据前',
                Value: 'BeforeModifyActions',
                HasInternalAction: false
            },
            {
                Index: 8,
                Name: '修改数据',
                Value: 'ModifyActions',
                HasInternalAction: true
            },
            {
                Index: 9,
                Name: '修改数据后',
                Value: 'AfterModifyActions',
                HasInternalAction: false
            },
            {
                Index: 10,
                Name: '变更集Mapping',
                Value: 'ChangesetMappingActions',
                HasInternalAction: true
            },
            {
                Index: 11,
                Name: '新增数据前',
                Value: 'BeforeCreateActions',
                HasInternalAction: false
            },
            {
                Index: 12,
                Name: '新增数据',
                Value: 'CreateActions',
                HasInternalAction: true
            },
            {
                Index: 13,
                Name: '新增数据后',
                Value: 'AfterCreateActions',
                HasInternalAction: false
            },
            {
                Index: 14,
                Name: '删除数据前',
                Value: 'BeforeDeleteActions',
                HasInternalAction: false
            },
            {
                Index: 15,
                Name: '删除数据',
                Value: 'DeleteActions',
                HasInternalAction: true
            },
            {
                Index: 16,
                Name: '删除数据后',
                Value: 'AfterDeleteActions',
                HasInternalAction: false
            },
            {
                Index: 17,
                Name: '批量删除数据前',
                Value: 'BeforeMultiDeleteActions',
                HasInternalAction: false
            },
            {
                Index: 18,
                Name: '批量删除数据',
                Value: 'MultiDeleteActions',
                HasInternalAction: true
            },
            {
                Index: 19,
                Name: '批量删除数据后',
                Value: 'AfterMultiDeleteActions',
                HasInternalAction: false
            },
            {
                Index: 20,
                Name: '保存数据前',
                Value: 'BeforeSaveActions',
                HasInternalAction: false
            },
            {
                Index: 21,
                Name: '数据反向Mapping',
                Value: 'DataReversalMappingActions',
                HasInternalAction: true
            },
            {
                Index: 22,
                Name: '保存数据后',
                Value: 'AfterSaveActions',
                HasInternalAction: false
            },
            {
                Index: 23,
                Name: '变更集反向Mapping',
                Value: 'ChangesetReversalMappingActions',
                HasInternalAction: true
            }
        ]
    }];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Action执行参数
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * Action执行参数
 * @abstract
 */
class ViewModelParameter {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const VMParameterMode = {
    IN: 'IN',
    OUT: 'OUT',
    INOUT: 'INOUT',
};
/** @type {?} */
const VMParameterModes = [
    { text: 'IN', value: VMParameterMode.IN },
    { text: 'OUT', value: VMParameterMode.OUT },
    { text: 'INOUT', value: VMParameterMode.INOUT }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * ViewModel的返回值实体类
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * Action执行参数
 */
class ViewModelReturnValue extends ViewModelParameter {
    constructor() {
        super(...arguments);
        this.ParamName = undefined;
        this.Mode = VMParameterMode.OUT;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ViewModelVoidReturnValue extends ViewModelReturnValue {
    constructor() {
        super(...arguments);
        this.Assembly = 'mscorlib.dll';
        this.ClassName = 'Void';
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * VM Action抽象类
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * VM Action抽象类
 * @abstract
 */
class ViewModelAction {
    constructor() {
        // this.ID = Guid.newGuid();
        this.ReturnValue = new ViewModelVoidReturnValue();
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * VO操作类型枚举
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
const ViewModelActionType = {
    /**
     * BE的业务操作
     */
    BEAction: 'BEAction',
    /**
     * VM 操作
     */
    VMAction: 'VMAction',
    /**
     * 自定义操作
     */
    Custom: 'Custom',
};
/** @type {?} */
const ViewModelActionTypes = [
    { text: 'BE操作', value: ViewModelActionType.BEAction },
    { text: 'VM操作', value: ViewModelActionType.VMAction },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * VM上映射的BE操作类
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * VM上映射的BE操作类
 */
class MappedBizAction extends ViewModelAction {
    constructor() {
        super(...arguments);
        /**
         * 类型
         */
        this.Type = ViewModelActionType.BEAction;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * VO上映射的构件操作类
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * VO上映射的构件操作类
 */
class MappedCdpAction extends ViewModelAction {
    constructor() {
        super();
        /**
         * 类型
         */
        this.Type = ViewModelActionType.VMAction;
        this.IsGenerateComponent = true;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Action执行结果
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * Action执行结果
 */
class VMActionResult {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 模板VO信息
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 模板VO信息
 */
class TemplateVoInfo {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Action执行参数
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * Action执行参数
 */
class MappedBizActionParameter extends ViewModelParameter {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Action执行参数
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * Action执行参数
 */
class MappedCdpActionParameter extends ViewModelParameter {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const VMParameterType = {
    /**
     * 字符型
     */
    String: 'String',
    /**
     * 布尔型
     */
    Boolean: 'Boolean',
    /**
     * 整数
     */
    Int32: 'Int32',
    /**
     * 浮点数
     */
    Decimal: 'Decimal',
    /**
     * 双浮点
     */
    Double: 'Double',
    /**
     * 时间
     */
    DateTime: 'DateTime',
    /**
     * 对象类型
     */
    Object: 'Object',
    /**
     * 自定义
     */
    Custom: 'Custom',
};
/** @type {?} */
const VMParameterTypes = [
    { index: 0, text: '字符串', value: VMParameterType.String, assembly: 'mscorlib.dll', className: 'System.String', javaClassName: 'java.lang.String' },
    { index: 1, text: '布尔型', value: VMParameterType.Boolean, assembly: 'mscorlib.dll', className: 'System.Boolean', javaClassName: 'java.lang.Boolean' },
    { index: 2, text: '整数', value: VMParameterType.Int32, assembly: 'mscorlib.dll', className: 'System.Int32', javaClassName: 'java.lang.Integer' },
    { index: 3, text: '浮点数', value: VMParameterType.Decimal, assembly: 'mscorlib.dll', className: 'System.Single', javaClassName: 'java.math.BigDecimal' },
    { index: 4, text: '双浮点数', value: VMParameterType.Double, assembly: 'mscorlib.dll', className: 'System.Double', javaClassName: 'java.lang.Double' },
    { index: 5, text: '时间类型', value: VMParameterType.DateTime, assembly: 'mscorlib.dll', className: 'System.DateTime', javaClassName: 'java.util.Date' },
    { index: 6, text: '对象类型', value: VMParameterType.Object, assembly: 'mscorlib.dll', className: 'System.Object', javaClassName: 'java.lang.Object' },
    { index: 7, text: '自定义类型', value: VMParameterType.Custom, assembly: '', className: '', javaClassName: '' },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const VMCollectionParameterType = {
    /**
     * 未使用集合形式
     */
    None: 'None',
    /**
     * 列表
     */
    List: 'List',
    /**
     * 数组
     */
    Array: 'Array',
};
/** @type {?} */
const VMCollectionParameterTypes = [
    { index: 0, text: '无', value: VMCollectionParameterType.None },
    { index: 1, text: '列表', value: VMCollectionParameterType.List },
    { index: 2, text: '数组', value: VMCollectionParameterType.Array }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 值帮助配置类
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 值帮助配置类
 */
class ValueHelpConfig {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * VO映射接口
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * VO映射接口
 */
class ViewModelMapping {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 帮助扩展动作（帮助前，帮助后等等一系列帮助的扩展）
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 帮助扩展动作（帮助前，帮助后等等一系列帮助的扩展）
 */
class HelpExtendAction {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * VO映射接口
 */
class GspVoElementMapping extends ViewModelMapping {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * VO映射接口
 */
class GspVoObjectMapping extends ViewModelMapping {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 权限控制类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
const ElementHelpType = {
    /**
     * 使用传递的权限，不限制权限，Action没有绑定的业务操作
     */
    TransAuth: 'TransAuth',
    /**
     * 使用传递的权限，并且限制在Action所绑定的业务操作内
     */
    TransAndInsideAuth: 'TransAndInsideAuth',
    /**
     * 不使用传递的权限，使用Action绑定的业务操作（只绑定一个业务操作）
     */
    MustHasAuth: 'MustHasAuth',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 映射类型枚举
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
const MappingType = {
    /**
     * BE
     */
    BizEntity: 'BizEntity',
    /**
     * DM
     */
    DataModel: 'DataModel',
    /**
     * BE节点
     */
    BEObject: 'BEObject',
    /**
     * GspDataModel节点
     */
    DMObject: 'DMObject',
    /**
     * 节点元素
     */
    Element: 'Element',
    /**
     * BE操作
     */
    BizOperation: 'BizOperation',
    /**
     * 构件方法
     */
    ComponentMethod: 'ComponentMethod',
    /**
     * 数据对象
     */
    DataObject: 'DataObject',
    /**
     * 数据对象列
     */
    GSPColumn: 'GSPColumn',
};
/** @type {?} */
const MappingTypes = [
    { text: '业务实体', value: MappingType.BizEntity },
    { text: '业务实体节点', value: MappingType.BEObject },
    { text: '业务实体字段', value: MappingType.Element },
    { text: '业务实体操作', value: MappingType.BizOperation },
    { text: '构件方法', value: MappingType.ComponentMethod },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 映射元数据类型枚举
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
const MetadataType = {
    /**
     * BE
     */
    BizEntity: 'BizEntity',
    /**
     * QO
     */
    QueryObject: 'QueryObject',
};
/** @type {?} */
const MetadataTypes = [
    { text: '业务实体', value: MetadataType.BizEntity },
    { text: '查询', value: MetadataType.QueryObject },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const GspVoObjectSourceType = {
    BeObject: 'BeObject',
    QoObject: 'QoObject',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const GspVoElementSourceType = {
    BeElement: 'BeElement',
    QoElement: 'QoElement',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { GSPViewModel, GSPViewModelElement, GSPViewObject, VoDataExtendInfo, VoDataExtendInfos, MappedBizAction, MappedCdpAction, ViewModelAction, VMActionResult, TemplateVoInfo, ViewModelParameter, MappedBizActionParameter, MappedCdpActionParameter, VMParameterMode, VMParameterModes, VMParameterType, VMParameterTypes, VMCollectionParameterType, VMCollectionParameterTypes, ViewModelReturnValue, ViewModelVoidReturnValue, ValueHelpConfig, ViewModelMapping, HelpExtendAction, GspVoElementMapping, GspVoObjectMapping, ElementHelpType, MappingType, MappingTypes, MetadataType, MetadataTypes, ViewModelActionType, ViewModelActionTypes, GspVoObjectSourceType, GspVoElementSourceType };
