// @farris/ui-filter/esm5/lib/components/singlenumbercomponent/number.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, EventEmitter, Output, ElementRef } from '@angular/core';
import { SolutionService } from '../../service/solution.service';
var SingleNumberComponent = /** @class */ (function () {
    function SingleNumberComponent(solutionService, element) {
        this.solutionService = solutionService;
        this.element = element;
        this.searchHandle = new EventEmitter();
        /**
         * 判断是否是在扩展区域展示
         */
        this.extend = false;
        /**
         * 判断是否显示label
         */
        this.showlabel = true;
        this.canNull = true;
    }
    Object.defineProperty(SingleNumberComponent.prototype, "isExtend", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.extend && this.autoLabel;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SingleNumberComponent.prototype, "hidenLabel", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.showlabel;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SingleNumberComponent.prototype, "item", {
        get: /**
         * @return {?}
         */
        function () {
            return this._item;
        },
        set: /**
         * @param {?} nr
         * @return {?}
         */
        function (nr) {
            this._item = nr;
            this.precision = nr.control.precision;
            this.bigNumber = nr.control.bigNumber;
            if (this._item && this._item.control && this._item.control.hasOwnProperty('isExtend')) {
                this.extend = this._item.control.isExtend;
                if (!this.extend) {
                    this.showlabel = this._item.control.showLabel;
                }
            }
            if (!this.bigNumber) {
                this.min = -2147483648;
                this.max = 2147483647;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    SingleNumberComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.item) {
            this.placeholder = this.item.placeHolder;
            // const type = Object.prototype.toString.call(this.item.placeHolder);
            // if(type == '[object Undefined]'){
            //     if(this.item.control && this.item.control.hasOwnProperty('placeholder')){
            //         this.placeholder = this.item.control.placeholder;
            //     }
            // }
            // else{
            //     this.placeholder = this.item.placeHolder;
            // }
        }
    };
    /**
     * @param {?} field
     * @return {?}
     */
    SingleNumberComponent.prototype.getConditionItemTitle = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        return this.solutionService.getConditionItemTitle(field);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    SingleNumberComponent.prototype.numberChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.item.value.value = event;
        if (this.disabled) {
            return;
        }
        this.searchHandle.emit();
    };
    SingleNumberComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-filter-single-number',
                    template: "<label class=\"col-form-label\" *ngIf=\"item && showlabel\" title=\"{{getConditionItemTitle(item)}}\">\r\n    <span class=\"farris-label-info text-danger\" *ngIf=\"item.control.required\">*</span>\r\n    <span class=\"farris-label-text\">{{item.fieldName}}</span>\r\n</label>\r\n<div class=\"farris-input-wrap d-flex\">\r\n    <!-- <farris-number-range class=\"f-utils-fill-flex-row\" [canNull]=\"canNull\" [disabled]=\"disabled\" [beginValue]=\"beginValue\" [endValue]=\"endValue\" (beginValueChange)=\"beginValueChange($event)\" (endValueChange)=\"endValueChange($event)\" (valueChange)=\"numberChange($event)\" [beginPlaceHolder]=\"beginPlaceHolder\" [endPlaceHolder]=\"endPlaceHolder\" [placeholder]=\"placeholder\" [precision]=\"precision\" [bigNumber]=\"bigNumber\" [min]=\"min\" [max]=\"max\"></farris-number-range> -->\r\n\r\n    <farris-number-spinner\r\n        class=\"f-utils-fill-flex-row\"\r\n        [(ngModel)]=\"item.value.value\"\r\n        [canNull]=\"canNull\"\r\n        [precision]=\"precision\"\r\n        [bigNumber]=\"bigNumber\"\r\n        [min]=\"min\" \r\n        [max]=\"max\"\r\n        [placeholder]=\"placeholder\"\r\n        (valueChange)=\"numberChange($event)\"\r\n    ></farris-number-spinner>\r\n\r\n</div>"
                }] }
    ];
    /** @nocollapse */
    SingleNumberComponent.ctorParameters = function () { return [
        { type: SolutionService },
        { type: ElementRef }
    ]; };
    SingleNumberComponent.propDecorators = {
        autoLabel: [{ type: Input }],
        disabled: [{ type: Input }],
        searchHandle: [{ type: Output, args: ['searchHandle',] }],
        isExtend: [{ type: HostBinding, args: ['class.filter-item-wrap',] }],
        hidenLabel: [{ type: HostBinding, args: ['class.filter-item-hiddenlabel',] }],
        item: [{ type: Input }]
    };
    return SingleNumberComponent;
}());
export { SingleNumberComponent };
if (false) {
    /** @type {?} */
    SingleNumberComponent.prototype.autoLabel;
    /** @type {?} */
    SingleNumberComponent.prototype.disabled;
    /** @type {?} */
    SingleNumberComponent.prototype.searchHandle;
    /**
     * 判断是否是在扩展区域展示
     * @type {?}
     */
    SingleNumberComponent.prototype.extend;
    /**
     * 判断是否显示label
     * @type {?}
     */
    SingleNumberComponent.prototype.showlabel;
    /** @type {?} */
    SingleNumberComponent.prototype.canNull;
    /** @type {?} */
    SingleNumberComponent.prototype.number;
    /** @type {?} */
    SingleNumberComponent.prototype.precision;
    /** @type {?} */
    SingleNumberComponent.prototype.placeholder;
    /** @type {?} */
    SingleNumberComponent.prototype.bigNumber;
    /** @type {?} */
    SingleNumberComponent.prototype.max;
    /** @type {?} */
    SingleNumberComponent.prototype.min;
    /** @type {?} */
    SingleNumberComponent.prototype._item;
    /** @type {?} */
    SingleNumberComponent.prototype.numberRange;
    /**
     * @type {?}
     * @private
     */
    SingleNumberComponent.prototype.solutionService;
    /**
     * @type {?}
     * @private
     */
    SingleNumberComponent.prototype.element;
}
