// @edp-pmf/bpmn-model/dist/cjs/lib/model/Association.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Association = void 0;
var tslib_1 = require("tslib");
var Artifact_1 = require("./Artifact");
var AssociationDirection_1 = require("./AssociationDirection");
var ExtensionMethod_1 = require("./ExtensionMethod");
var Association = /** @class */ (function (_super) {
    tslib_1.__extends(Association, _super);
    function Association(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.Association";
        _this.AssociationDirection = AssociationDirection_1.AssociationDirection.NONE;
        return _this;
    }
    Association.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "SourceRef", this.SourceRef);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "TargetRef", this.TargetRef);
        return obj;
    };
    Association.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.SourceRef = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "SourceRef");
        this.TargetRef = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "TargetRef");
    };
    Association.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
    };
    Association.prototype.NewInstance = function () {
        return new Association(this.Model);
    };
    return Association;
}(Artifact_1.Artifact));
exports.Association = Association;
