// @edp-pmf/bpmn-model/dist/cjs/lib/model/CatchEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CatchEvent = void 0;
var tslib_1 = require("tslib");
var BpmnModelHelper_1 = require("./BpmnModelHelper");
var DataOutput_1 = require("./DataOutput");
var DataOutputAssociation_1 = require("./DataOutputAssociation");
var ValidateHelper_1 = require("./ValidateHelper");
var Event_1 = require("./Event");
var CatchEvent = /** @class */ (function (_super) {
    tslib_1.__extends(CatchEvent, _super);
    function CatchEvent(model) {
        var _this = _super.call(this, model) || this;
        _this.dataOutputAssociationDic = {};
        _this.dataOutputDIc = {};
        return _this;
    }
    CatchEvent.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.dataOutputAssociationDic = {};
        this.dataOutputDIc = {};
    };
    CatchEvent.prototype.AddDataOutput = function (item) {
        this.CheckVariableElement(item);
        this.dataOutputDIc[item.Id] = item;
        this.Model.AddHasIdElement(item);
    };
    CatchEvent.prototype.GetDataOutput = function (dataOutputId) {
        if (dataOutputId == null)
            return null;
        if (this.dataOutputDIc.hasOwnProperty("dataOutputId"))
            return this.dataOutputDIc[dataOutputId];
        return null;
    };
    CatchEvent.prototype.GetDataOutputs = function () {
        var dataOutputs = new Array();
        for (var item in this.dataOutputDIc) {
            dataOutputs.push(this.dataOutputDIc[item]);
        }
        return dataOutputs;
    };
    CatchEvent.prototype.RemoveDataOutput = function (dataOutputId) {
        delete this.dataOutputDIc[dataOutputId];
        this.Model.RemoveHasIdElement(dataOutputId);
    };
    CatchEvent.prototype.AddDataOutputAssociation = function (association) {
        ValidateHelper_1.ValidateHelper.CheckHasIdElement(association);
        this.dataOutputAssociationDic[association.Id] = association;
        this.Model.AddHasIdElement(association);
    };
    CatchEvent.prototype.GetDataOutputAssociation = function (associationId) {
        if (associationId == null)
            return null;
        if (this.dataOutputAssociationDic.hasOwnProperty("associationId"))
            return this.dataOutputAssociationDic[associationId];
        return null;
    };
    CatchEvent.prototype.GetDataOutputAssociations = function () {
        var dataOutputAssociations = new Array();
        for (var item in this.dataOutputAssociationDic) {
            dataOutputAssociations.push(this.dataOutputAssociationDic[item]);
        }
        return dataOutputAssociations;
    };
    CatchEvent.prototype.RemoveDataOutputAssociation = function (associationId) {
        delete this.dataOutputAssociationDic[associationId];
        this.Model.RemoveHasIdElement(associationId);
    };
    CatchEvent.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        var dataOutputs = new Array();
        for (var item in this.dataOutputDIc) {
            var element = this.dataOutputDIc[item].ConvertToJsonObject();
            dataOutputs.push(element);
        }
        if (dataOutputs.length > 0) {
            BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "dataOutputs", dataOutputs);
        }
        var dataOutputAssociations = new Array();
        for (var item in this.dataOutputAssociationDic) {
            var element = this.dataOutputAssociationDic[item].ConvertToJsonObject();
            dataOutputAssociations.push(element);
        }
        if (dataOutputAssociations.length > 0) {
            BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "dataOutputAssociations", dataOutputAssociations);
        }
        return obj;
    };
    CatchEvent.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        if (obj.hasOwnProperty("dataOutputs") && Array.isArray(obj.dataOutputs)) {
            for (var _i = 0, _a = obj.dataOutputs; _i < _a.length; _i++) {
                var item = _a[_i];
                var dataOutput = new DataOutput_1.DataOutput(this.Model);
                dataOutput.LoadFromJsonObject(item);
                this.AddDataOutput(dataOutput);
            }
        }
        if (obj.hasOwnProperty("dataOutputAssociations") && Array.isArray(obj.dataOutputAssociations)) {
            for (var item in obj.dataOutputAssociations) {
                var dataOutputAssociation = new DataOutputAssociation_1.DataOutputAssociation(this.Model);
                dataOutputAssociation.LoadFromJsonObject(item);
                this.AddDataOutputAssociation(dataOutputAssociation);
            }
        }
    };
    CatchEvent.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
        for (var item in this.dataOutputDIc) {
            this.dataOutputDIc[item].Validate(errors);
        }
        for (var item in this.dataOutputAssociationDic) {
            this.dataOutputAssociationDic[item].Validate(errors);
        }
    };
    CatchEvent.prototype.CheckVariableElement = function (variableElement) {
        ValidateHelper_1.ValidateHelper.CheckVariableElement(variableElement, null, this, null);
    };
    return CatchEvent;
}(Event_1.Event));
exports.CatchEvent = CatchEvent;
