// @edp-pmf/bpmn-model/dist/cjs/lib/model/DataAssociation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAssociation = void 0;
var tslib_1 = require("tslib");
var BaseElement_1 = require("./BaseElement");
var BpmnModelHelper_1 = require("./BpmnModelHelper");
var FormalExpression_1 = require("./FormalExpression");
var ValidationError_1 = require("./ValidationError");
var ValidationLevel_1 = require("./ValidationLevel");
var DataAssociation = /** @class */ (function (_super) {
    tslib_1.__extends(DataAssociation, _super);
    function DataAssociation(model) {
        return _super.call(this, model) || this;
    }
    DataAssociation.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.TargetRef = null;
        this.Transformation = null;
    };
    DataAssociation.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "TargetRef", this.TargetRef);
        if (this.Transformation == null) {
            throw new Error("Transformation is null.");
        }
        else {
            BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Transformation", this.Transformation.ConvertToJsonObject());
        }
        return obj;
    };
    DataAssociation.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.TargetRef = obj.TargetRef;
        if (obj.hasOwnProperty("Transformation")) {
            this.Transformation = new FormalExpression_1.FormalExpression(this.Model);
            this.Transformation.LoadFromJsonObject(obj.Transformation);
        }
    };
    DataAssociation.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
        if (this.Transformation != null) {
            this.Transformation.Validate(errors);
        }
        else {
            var error = new ValidationError_1.ValidationError();
            error.Element = this;
            error.ErrorLevel = ValidationLevel_1.ValidationLevel.$_Error;
            error.ErrorCode = "";
            error.ErrorMessage = "没有设定赋值表达式";
            errors.push(error);
        }
        if (this.TargetRef == null) {
            var error = new ValidationError_1.ValidationError();
            error.Element = this;
            error.ErrorLevel = ValidationLevel_1.ValidationLevel.$_Error;
            error.ErrorCode = "";
            error.ErrorMessage = "没有设定关联的目标";
            errors.push(error);
        }
    };
    return DataAssociation;
}(BaseElement_1.BaseElement));
exports.DataAssociation = DataAssociation;
