// @edp-pmf/bpmn-model/dist/cjs/lib/model/DataObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataObject = void 0;
var tslib_1 = require("tslib");
var BpmnModelHelper_1 = require("./BpmnModelHelper");
var FlowElement_1 = require("./FlowElement");
var DataType_1 = require("./mex/DataType");
var ValidationError_1 = require("./ValidationError");
var ValidationLevel_1 = require("./ValidationLevel");
var DataObject = /** @class */ (function (_super) {
    tslib_1.__extends(DataObject, _super);
    function DataObject(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.DataObject";
        _this.MEXTAG = "bpmn.DataObject";
        _this.IsAllowNull = true;
        _this.IsCollection = false;
        return _this;
    }
    DataObject.prototype.clone = function () {
        var clone = _super.prototype.clone.call(this);
        clone.DataType = this.DataType;
        clone.IsAllowNull = this.IsAllowNull;
        clone.IsCollection = this.IsCollection;
        return clone;
    };
    DataObject.prototype.NewInstance = function () {
        return new DataObject(this.Model);
    };
    DataObject.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.DataType = null;
        this.IsAllowNull = true;
        this.IsCollection = false;
    };
    DataObject.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "IsAllowNull", this.IsAllowNull);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "IsCollection", this.IsCollection);
        return obj;
    };
    DataObject.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.IsAllowNull = obj.IsAllowNull;
        this.IsCollection = obj.IsCollection;
    };
    DataObject.prototype.GetExtensionElements4ConvertTo = function () {
        var list = _super.prototype.GetExtensionElements4ConvertTo.call(this);
        if (this.DataType != null) {
            this.DataType.MexTag = this.MEXTAG;
            list.push(this.DataType);
        }
        return list;
    };
    DataObject.prototype.SetExtensionElements4LoadFrom = function (extensionElements) {
        _super.prototype.SetExtensionElements4LoadFrom.call(this, extensionElements);
        for (var _i = 0, extensionElements_1 = extensionElements; _i < extensionElements_1.length; _i++) {
            var extensionElement = extensionElements_1[_i];
            if (extensionElement instanceof DataType_1.DataType && extensionElement.MexTag == this.MEXTAG) {
                this.DataType = extensionElement;
            }
        }
    };
    DataObject.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
        if (this.DataType != null) {
            this.DataType.Validate(errors);
        }
        else {
            var error = new ValidationError_1.ValidationError();
            error.Element = this;
            error.ErrorLevel = ValidationLevel_1.ValidationLevel.$_Error;
            error.ErrorCode = "";
            error.ErrorMessage = "没有设定变量类型";
            errors.push(error);
        }
    };
    return DataObject;
}(FlowElement_1.FlowElement));
exports.DataObject = DataObject;
