// @edp-pmf/bpmn-model/dist/cjs/lib/model/DataInput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataInput = void 0;
var tslib_1 = require("tslib");
var BaseElement_1 = require("./BaseElement");
var BpmnModelHelper_1 = require("./BpmnModelHelper");
var DataType_1 = require("./mex/DataType");
var ValidationError_1 = require("./ValidationError");
var ValidationLevel_1 = require("./ValidationLevel");
var DataInput = /** @class */ (function (_super) {
    tslib_1.__extends(DataInput, _super);
    function DataInput(model) {
        var _this = _super.call(this, model) || this;
        _this.MEXTAG = "bpmn.DataInput";
        _this.IsAllowNull = true;
        _this.IsRequired = true;
        _this.IsCollection = false;
        return _this;
    }
    DataInput.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.IsAllowNull = true;
        this.IsRequired = true;
        this.Name = null;
        this.DataType = null;
        this.IsCollection = false;
    };
    DataInput.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Name", this.Name);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "IsAllowNull", this.IsAllowNull);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "IsRequired", this.IsRequired);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "IsCollection", this.IsCollection);
        return obj;
    };
    DataInput.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Name = obj.Name;
        this.IsAllowNull = obj.IsAllowNull;
        this.IsRequired = obj.IsRequired;
        this.IsCollection = obj.IsCollection;
    };
    DataInput.prototype.GetExtensionElements4ConvertTo = function () {
        var list = _super.prototype.GetExtensionElements4ConvertTo.call(this);
        if (this.DataType != null) {
            this.DataType.MexTag = this.MEXTAG;
            list.push(this.DataType);
        }
        return list;
    };
    DataInput.prototype.SetExtensionElements4LoadFrom = function (extensionElements) {
        _super.prototype.SetExtensionElements4LoadFrom.call(this, extensionElements);
        for (var _i = 0, extensionElements_1 = extensionElements; _i < extensionElements_1.length; _i++) {
            var extensionElement = extensionElements_1[_i];
            if (extensionElement instanceof DataType_1.DataType && extensionElement.MexTag == this.MEXTAG) {
                this.DataType = extensionElement;
            }
        }
    };
    DataInput.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
        if (this.DataType != null) {
            this.DataType.Validate(errors);
        }
        else {
            var error = new ValidationError_1.ValidationError();
            error.Element = this;
            error.ErrorLevel = ValidationLevel_1.ValidationLevel.$_Error;
            error.ErrorCode = "";
            error.ErrorMessage = "没有设定变量类型";
            errors.push(error);
        }
    };
    return DataInput;
}(BaseElement_1.BaseElement));
exports.DataInput = DataInput;
