// @edp-pmf/bpmn-model/dist/cjs/lib/model/ExclusiveGateway.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExclusiveGateway = void 0;
var tslib_1 = require("tslib");
var BpmnModelHelper_1 = require("./BpmnModelHelper");
var Gateway_1 = require("./Gateway");
var GatewayType_1 = require("./GatewayType");
var ExclusiveGateway = /** @class */ (function (_super) {
    tslib_1.__extends(ExclusiveGateway, _super);
    function ExclusiveGateway(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.ExclusiveGateway";
        _this.IncomingGatewayType = GatewayType_1.GatewayType.Exclusive;
        _this.OutgoingGatewayType = GatewayType_1.GatewayType.Exclusive;
        return _this;
    }
    ExclusiveGateway.prototype.NewInstance = function () {
        return new ExclusiveGateway(this.Model);
    };
    ExclusiveGateway.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Default = null;
    };
    ExclusiveGateway.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Default", this.Default);
        return obj;
    };
    ExclusiveGateway.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Default = obj.Default;
    };
    return ExclusiveGateway;
}(Gateway_1.Gateway));
exports.ExclusiveGateway = ExclusiveGateway;
