// @edp-pmf/bpmn-model/dist/cjs/lib/model/Expression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Expression = void 0;
var tslib_1 = require("tslib");
var BaseElement_1 = require("./BaseElement");
var MexExpression_1 = require("./mex/MexExpression");
var Expression = /** @class */ (function (_super) {
    tslib_1.__extends(Expression, _super);
    function Expression(model) {
        var _this = _super.call(this, model) || this;
        _this.MEXTAG = "bpmn.Expression";
        return _this;
    }
    Expression.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.MexExpression = null;
    };
    Expression.prototype.GetExtensionElements4ConvertTo = function () {
        var list = _super.prototype.GetExtensionElements4ConvertTo.call(this);
        if (this.MexExpression != null) {
            this.MexExpression.MexTag = this.MEXTAG;
            list.push(this.MexExpression);
        }
        return list;
    };
    Expression.prototype.SetExtensionElements4LoadFrom = function (extensionElements) {
        _super.prototype.SetExtensionElements4LoadFrom.call(this, extensionElements);
        for (var _i = 0, extensionElements_1 = extensionElements; _i < extensionElements_1.length; _i++) {
            var extensionElement = extensionElements_1[_i];
            if (extensionElement instanceof MexExpression_1.MexExpression && extensionElement.MexTag == this.MEXTAG) {
                this.MexExpression = extensionElement;
            }
        }
    };
    Expression.prototype.Validate = function (errors) {
        _super.prototype.Validate.call(this, errors);
        if (this.MexExpression != null) {
            this.MexExpression.Validate(errors);
        }
    };
    Expression.prototype.NewInstance = function () {
        return new Expression(this.Model);
    };
    Expression.prototype.clone = function () {
        var expression = new Expression(this.Model);
        if (this.MexExpression) {
            expression.MexExpression = this.MexExpression.clone();
        }
        return expression;
    };
    return Expression;
}(BaseElement_1.BaseElement));
exports.Expression = Expression;
