// @edp-pmf/bpmn-model/dist/cjs/lib/model/InclusiveGateway.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InclusiveGateway = void 0;
var tslib_1 = require("tslib");
var BpmnModelHelper_1 = require("./BpmnModelHelper");
var Gateway_1 = require("./Gateway");
var GatewayType_1 = require("./GatewayType");
var InclusiveGateway = /** @class */ (function (_super) {
    tslib_1.__extends(InclusiveGateway, _super);
    function InclusiveGateway(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.InclusiveGateway";
        _this.IncomingGatewayType = GatewayType_1.GatewayType.Inclusive;
        _this.OutgoingGatewayType = GatewayType_1.GatewayType.Inclusive;
        return _this;
    }
    InclusiveGateway.prototype.NewInstance = function () {
        return new InclusiveGateway(this.Model);
    };
    InclusiveGateway.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Default = null;
    };
    InclusiveGateway.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Default", this.Default);
        return obj;
    };
    InclusiveGateway.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Default = obj.Default;
    };
    return InclusiveGateway;
}(Gateway_1.Gateway));
exports.InclusiveGateway = InclusiveGateway;
