// @edp-pmf/bpmn-model/dist/cjs/lib/model/MultiInstanceLoopCharacteristics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiInstanceLoopCharacteristics = void 0;
var tslib_1 = require("tslib");
var Expression_1 = require("./Expression");
var LoopCharacteristics_1 = require("./LoopCharacteristics");
var ExtensionCompletionConditionRuler_1 = require("./mex/ExtensionCompletionConditionRuler");
var ExtensionMethod_1 = require("./ExtensionMethod");
var MultiInstanceLoopCharacteristics = /** @class */ (function (_super) {
    tslib_1.__extends(MultiInstanceLoopCharacteristics, _super);
    function MultiInstanceLoopCharacteristics(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = MultiInstanceLoopCharacteristics.Clr_Type_ID;
        _this.completionConditionRules = new Array();
        _this.MEXTAG = "bpmn.MultiInstanceLoopCharacteristics";
        return _this;
    }
    MultiInstanceLoopCharacteristics.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.loopCardinality = null;
        this.completionCondition = null;
        this.completionConditionRules = new Array();
    };
    MultiInstanceLoopCharacteristics.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "Enable", this.enable);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "IsSequential", this.sequential);
        if (this.loopCardinality != null) {
            ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "LoopCardinality", this.loopCardinality.ConvertToJsonObject());
        }
        if (this.completionCondition != null) {
            ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "CompletionCondition", this.completionCondition.ConvertToJsonObject());
        }
        return obj;
    };
    MultiInstanceLoopCharacteristics.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.enable = obj.Enable;
        this.sequential = obj.IsSequential;
        if (obj.hasOwnProperty("LoopCardinality")) {
            this.loopCardinality = new Expression_1.Expression(this.Model);
            this.loopCardinality.LoadFromJsonObject(obj.LoopCardinality);
        }
        if (obj.hasOwnProperty("CompletionCondition")) {
            this.completionCondition = new Expression_1.Expression(this.Model);
            this.completionCondition.LoadFromJsonObject(obj.CompletionCondition);
        }
    };
    MultiInstanceLoopCharacteristics.prototype.GetExtensionElements4ConvertTo = function () {
        var list = _super.prototype.GetExtensionElements4ConvertTo.call(this);
        if (this.completionConditionRules) {
            for (var _i = 0, _a = this.completionConditionRules; _i < _a.length; _i++) {
                var item = _a[_i];
                item.MexTag = this.MEXTAG;
                list.push(item);
            }
        }
        return list;
    };
    MultiInstanceLoopCharacteristics.prototype.SetExtensionElements4LoadFrom = function (extensionElements) {
        _super.prototype.SetExtensionElements4LoadFrom.call(this, extensionElements);
        for (var _i = 0, extensionElements_1 = extensionElements; _i < extensionElements_1.length; _i++) {
            var extensionElement = extensionElements_1[_i];
            if (extensionElement instanceof ExtensionCompletionConditionRuler_1.ExtensionCompletionConditionRuler && extensionElement.MexTag == this.MEXTAG) {
                this.completionConditionRules.push(extensionElement);
            }
        }
    };
    MultiInstanceLoopCharacteristics.prototype.NewInstance = function () {
        return new MultiInstanceLoopCharacteristics(this.Model);
    };
    MultiInstanceLoopCharacteristics.prototype.clone = function () {
        var loopCharacteristics = new MultiInstanceLoopCharacteristics(this.Model);
        loopCharacteristics.enable = this.enable;
        loopCharacteristics.sequential = this.sequential;
        if (this.loopCardinality) {
            loopCharacteristics.loopCardinality = this.loopCardinality.clone();
        }
        loopCharacteristics.completionCondition = this.completionCondition;
        loopCharacteristics.completionConditionRules = new Array();
        if (this.completionConditionRules) {
            this.completionConditionRules.forEach(function (com) { return loopCharacteristics.completionConditionRules.push(com.clone()); });
        }
        // loopCharacteristics.completionConditionRules = this.completionConditionRules;
        return loopCharacteristics;
    };
    MultiInstanceLoopCharacteristics.Clr_Type_ID = "wf.MultiInstanceLoopCharacteristics";
    MultiInstanceLoopCharacteristics.MEXTAG = "wf.CallActivitySubProcess";
    return MultiInstanceLoopCharacteristics;
}(LoopCharacteristics_1.LoopCharacteristics));
exports.MultiInstanceLoopCharacteristics = MultiInstanceLoopCharacteristics;
