// @edp-pmf/bpmn-model/dist/cjs/lib/model/ResourceRole.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceRole = void 0;
var tslib_1 = require("tslib");
var BaseElement_1 = require("./BaseElement");
var BpmnModelHelper_1 = require("./BpmnModelHelper");
var ResourceAssignmentExpression_1 = require("./ResourceAssignmentExpression");
var ResourceRole = /** @class */ (function (_super) {
    tslib_1.__extends(ResourceRole, _super);
    function ResourceRole(model) {
        return _super.call(this, model) || this;
    }
    ResourceRole.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Name = null;
        this.ResourceAssignmentExpression = null;
    };
    ResourceRole.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "Name", this.Name);
        BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "ClrTypeID", this.ClrTypeID);
        if (this.ResourceAssignmentExpression != null) {
            BpmnModelHelper_1.BpmnModelHelper.SetAttributeValue(obj, "ResourceAssignmentExpression", this.ResourceAssignmentExpression.ConvertToJsonObject());
        }
        return obj;
    };
    ResourceRole.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Name = obj.Name;
        if (obj.hasOwnProperty("ResourceAssignmentExpression")) {
            this.ResourceAssignmentExpression = new ResourceAssignmentExpression_1.ResourceAssignmentExpression(this.Model);
            this.ResourceAssignmentExpression.LoadFromJsonObject(obj.ResourceAssignmentExpression);
        }
    };
    return ResourceRole;
}(BaseElement_1.BaseElement));
exports.ResourceRole = ResourceRole;
