// @edp-pmf/bpmn-model/dist/cjs/lib/model/Signal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Signal = void 0;
var tslib_1 = require("tslib");
var CatchEvent_1 = require("./CatchEvent");
var ExtensionMethod_1 = require("./ExtensionMethod");
var Signal = /** @class */ (function (_super) {
    tslib_1.__extends(Signal, _super);
    function Signal(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.Signal";
        return _this;
    }
    Signal.prototype.convertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "ClrTypeID", this.ClrTypeID);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "Name", this.Name);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "Scope", this.Scope);
        return obj;
    };
    Signal.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Name = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "Name");
        this.Scope = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "Scope");
    };
    Signal.prototype.clone = function () {
        var clone = _super.prototype.clone.call(this);
        clone.Name = this.Name;
        clone.Scope = this.Scope;
        return clone;
    };
    Signal.prototype.NewInstance = function () {
        return new Signal(this.Model);
    };
    return Signal;
}(CatchEvent_1.CatchEvent));
exports.Signal = Signal;
