// @edp-pmf/bpmn-model/dist/cjs/lib/model/SignalEventDefinition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignalEventDefinition = void 0;
var tslib_1 = require("tslib");
var EventDefinition_1 = require("./EventDefinition");
var ExtensionMethod_1 = require("./ExtensionMethod");
var SignalEventDefinition = /** @class */ (function (_super) {
    tslib_1.__extends(SignalEventDefinition, _super);
    function SignalEventDefinition(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.SignalEventDefinition";
        return _this;
    }
    SignalEventDefinition.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "ClrTypeID", this.ClrTypeID);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "SignalRef", this.SignalRef);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "SignalExpression", this.SignalExpression);
        ExtensionMethod_1.ExtensionMethod.SetAttributeValue(obj, "Async", this.Async);
        return obj;
    };
    SignalEventDefinition.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.SignalRef = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "SignalRef");
        this.SignalExpression = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4String(obj, "SignalExpression");
        this.Async = ExtensionMethod_1.ExtensionMethod.GetAttributeValue4Boolean(obj, "Async");
    };
    SignalEventDefinition.prototype.NewInstance = function () {
        var clone = new SignalEventDefinition(this.Model);
        clone.SignalRef = this.SignalRef;
        clone.SignalExpression = this.SignalExpression;
        clone.Async = this.Async;
        return clone;
    };
    return SignalEventDefinition;
}(EventDefinition_1.EventDefinition));
exports.SignalEventDefinition = SignalEventDefinition;
