// @edp-pmf/bpmn-model/dist/cjs/lib/model/SplitActivity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SplitActivity = void 0;
var tslib_1 = require("tslib");
var ServiceTask_1 = require("./ServiceTask");
var SplitActivity = /** @class */ (function (_super) {
    tslib_1.__extends(SplitActivity, _super);
    function SplitActivity(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = SplitActivity.Clr_Type_ID;
        return _this;
    }
    SplitActivity.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
    };
    SplitActivity.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        return obj;
    };
    SplitActivity.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
    };
    SplitActivity.prototype.NewInstance = function () {
        return new SplitActivity(this.Model);
    };
    SplitActivity.prototype.clone = function () {
        var c = _super.prototype.clone.call(this);
        c.Name = this.Name;
        c.NameLanguage = this.NameLanguage && tslib_1.__assign({}, this.NameLanguage);
        return c;
    };
    SplitActivity.Clr_Type_ID = "bpmn.SplitActivity";
    SplitActivity.MEXTAG = 'bpmn.SplitActivity';
    return SplitActivity;
}(ServiceTask_1.ServiceTask));
exports.SplitActivity = SplitActivity;
