// @gsp-aif/runtime-api/esm2015/lib/aif-push-select-creation-rule/aif-push-select-creation-rule/aif-push-select-creation-rule.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from "@angular/core";
import { Subject } from "rxjs";
import { AifPushSelectCreationRuleInternalService } from '../service/aif-push-select-creation-rule.internal.service';
export class AifPushSelectCreationRuleComponent {
    /**
     * @param {?} crService
     */
    constructor(crService) {
        this.crService = crService;
        this.voTypeCodes = [];
        this.voTypeObjs = {};
        this.getCreationRuleResultEntities = {};
        this.selectedVoTypeCode = null;
        this.selectedGetCreationRuleResultEntity = null;
        this.bsModalRef = null;
        this.subject = new Subject();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} args
     * @return {?}
     */
    getCreationRule(args) {
        this.crService.getCreationRule(args).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            console.log(data);
            for (let item of data) {
                if (!this.voTypeCodes.includes(item.TargetVoType)) {
                    this.voTypeCodes.push(item.TargetVoType);
                    this.getCreationRuleResultEntities[item.TargetVoType] = [];
                    this.voTypeObjs[item.TargetVoType] = {
                        "VoType": item.TargetVoType,
                        "VoName": item.TargetVoProxyName
                    };
                }
                this.getCreationRuleResultEntities[item.TargetVoType].push(item);
            }
            if (data.length > 0) {
                this.selectedVoTypeCode = this.voTypeCodes[0];
                this.selectedGetCreationRuleResultEntity = data[0];
            }
        }));
    }
    /**
     * @param {?} beTypeCode
     * @return {?}
     */
    changeVoTypeRadio(beTypeCode) {
        this.selectedVoTypeCode = beTypeCode;
        this.selectedGetCreationRuleResultEntity = !this.isNullOrUndefined(this.getCreationRuleResultEntities[this.selectedVoTypeCode]) && this.getCreationRuleResultEntities[this.selectedVoTypeCode].length > 0 ? this.getCreationRuleResultEntities[this.selectedVoTypeCode][0] : null;
    }
    /**
     * @return {?}
     */
    clickOk() {
        this.subject.next((/** @type {?} */ (this.selectedGetCreationRuleResultEntity)));
        this.subject.complete();
        this.bsModalRef.close();
    }
    /**
     * @return {?}
     */
    clickClose() {
        this.bsModalRef.close();
    }
    /**
     * @private
     * @param {?} arg
     * @return {?}
     */
    isNullOrUndefined(arg) {
        return arg == null;
    }
}
AifPushSelectCreationRuleComponent.decorators = [
    { type: Component, args: [{
                selector: "app-aif-push-select-creation-rule",
                template: "<div style=\"height:100%;\">\r\n    <kendo-splitter orientation=\"vertical\">\r\n        <kendo-splitter-pane [resizable]=\"false\" [scrollable]=\"false\" [collapsible]=\"false\" size=\"40%\"\r\n            class=\"aif-splitter-pane-padding\">\r\n            <div>\r\n                <div>\u9009\u62E9\u76EE\u6807\u5355\u636E\u7C7B\u578B</div>\r\n                <div style=\"height:8rem;overflow-y: auto;\">\r\n                    <label *ngFor=\"let voTypeCode of voTypeCodes;let i=index\" class=\"col-sm-12 col-md-5\"\r\n                        style=\"padding:0 0;\">\r\n                        <input type=\"radio\" name=\"voTypeCode\" class=\"\" [checked]=\"i==0\"\r\n                            [value]=\"voTypeObjs[voTypeCode].VoType\"\r\n                            (click)=\"changeVoTypeRadio(voTypeCode)\">{{voTypeObjs[voTypeCode].VoName}}\r\n                    </label>\r\n                </div>\r\n            </div>\r\n        </kendo-splitter-pane>\r\n\r\n        <kendo-splitter-pane [resizable]=\"false\" [scrollable]=\"false\" [collapsible]=\"false\" size=\"40%\"\r\n            class=\"aif-splitter-pane-padding\">\r\n            <div>\r\n                <div>\u9009\u62E9\u751F\u5355\u89C4\u5219</div>\r\n                <kendo-dropdownlist class=\"form-control\" [data]=\"getCreationRuleResultEntities[selectedVoTypeCode]\"\r\n                    textField=\"Name\" valueField=\"Id\" [(ngModel)]=\"selectedGetCreationRuleResultEntity\">\r\n                </kendo-dropdownlist>\r\n            </div>\r\n        </kendo-splitter-pane>\r\n\r\n        <kendo-splitter-pane [resizable]=\"false\" [scrollable]=\"false\" [collapsible]=\"false\" size=\"20%\"\r\n            class=\"aif-splitter-pane-padding\">\r\n            <div>\r\n                <div>\u63CF\u8FF0</div>\r\n                <textarea kendoTextArea readonly class=\"form-control\" name=\"description\" id=\"\" rows=\"3\"\r\n                    [ngModel]=\"selectedGetCreationRuleResultEntity&&selectedGetCreationRuleResultEntity.Description\"></textarea>\r\n            </div>\r\n        </kendo-splitter-pane>\r\n\r\n        <kendo-splitter-pane [resizable]=\"false\" [scrollable]=\"false\" [collapsible]=\"false\"\r\n            class=\"aif-splitter-pane-padding\">\r\n            <div style=\"float:right;\">\r\n                <button class=\"btn-primary btn btn-sm ml-2\" (click)=\"clickOk()\"\r\n                    [disabled]=\"!selectedGetCreationRuleResultEntity\">\u786E\u5B9A</button>\r\n                <button class=\"btn btn-sm ml-2 btn-default\" (click)=\"clickClose()\">\u53D6\u6D88</button>\r\n            </div>\r\n        </kendo-splitter-pane>\r\n\r\n    </kendo-splitter>\r\n</div>",
                styles: [".aif-splitter-pane-padding{padding:.5rem 1rem}"]
            }] }
];
/** @nocollapse */
AifPushSelectCreationRuleComponent.ctorParameters = () => [
    { type: AifPushSelectCreationRuleInternalService }
];
if (false) {
    /** @type {?} */
    AifPushSelectCreationRuleComponent.prototype.voTypeCodes;
    /** @type {?} */
    AifPushSelectCreationRuleComponent.prototype.voTypeObjs;
    /** @type {?} */
    AifPushSelectCreationRuleComponent.prototype.getCreationRuleResultEntities;
    /** @type {?} */
    AifPushSelectCreationRuleComponent.prototype.selectedVoTypeCode;
    /** @type {?} */
    AifPushSelectCreationRuleComponent.prototype.selectedGetCreationRuleResultEntity;
    /** @type {?} */
    AifPushSelectCreationRuleComponent.prototype.bsModalRef;
    /** @type {?} */
    AifPushSelectCreationRuleComponent.prototype.subject;
    /**
     * @type {?}
     * @private
     */
    AifPushSelectCreationRuleComponent.prototype.crService;
}
