// @farris/ui-language-textbox/esm2015/lib/language-textbox.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class LanguageTextboxService {
    constructor() {
        this.activePane = null;
        this.hide$ = new Subject();
    }
    /**
     * @param {?} p
     * @return {?}
     */
    setActivePane(p) {
        if (this.activePane && this.activePane != p) {
            this.activePane.hideDropDownPanel();
            this.activePane = null;
        }
        this.activePane = p;
    }
    /**
     * @return {?}
     */
    hide() {
        this.hide$.next();
    }
}
LanguageTextboxService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
LanguageTextboxService.ctorParameters = () => [];
/** @nocollapse */ LanguageTextboxService.ngInjectableDef = i0.defineInjectable({ factory: function LanguageTextboxService_Factory() { return new LanguageTextboxService(); }, token: LanguageTextboxService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    LanguageTextboxService.prototype.activePane;
    /** @type {?} */
    LanguageTextboxService.prototype.hide$;
}
