// @gsp-bef/gsp-cm-metadata/fesm5/gsp-bef-gsp-cm-metadata.js
import { __extends } from 'tslib';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CommonModelUtil = /** @class */ (function () {
    function CommonModelUtil() {
    }
    /**
     * @param {?} cm
     * @return {?}
     */
    CommonModelUtil.GetGeneratedConfigID = /**
     * @param {?} cm
     * @return {?}
     */
    function (cm) {
        if (cm.IsUseNamespaceConfig) {
            return cm.GeneratingAssembly + '.' + cm.Code;
        }
        return cm.Code;
    };
    /**
     * @private
     * @param {?} obj
     * @param {?} objId
     * @return {?}
     */
    CommonModelUtil.findObjectFromObjectByID = /**
     * @private
     * @param {?} obj
     * @param {?} objId
     * @return {?}
     */
    function (obj, objId) {
        /** @type {?} */
        var result = null;
        if (!obj) {
            return null;
        }
        if (obj.ID === objId) {
            return obj;
        }
        if (!obj.ContainChildObjects) {
            return null;
        }
        // 递归查找子节点
        if (obj.ContainChildObjects.length > 0) {
            for (var i = 0; i < obj.ContainChildObjects.length; i++) {
                if (obj.ContainChildObjects[i].ID === objId) {
                    result = obj.ContainChildObjects[i];
                    return result;
                }
                result = this.findObjectFromObjectByID(obj.ContainChildObjects[i], objId);
                if (result != null) {
                    return result;
                }
            }
        }
        return result;
    };
    /**
     * 根据objId查找对象
     * @param objId
     */
    /**
     * 根据objId查找对象
     * @param {?} cm
     * @param {?} objId
     * @return {?}
     */
    CommonModelUtil.prototype.findObjectById = /**
     * 根据objId查找对象
     * @param {?} cm
     * @param {?} objId
     * @return {?}
     */
    function (cm, objId) {
        /** @type {?} */
        var result = null;
        result = CommonModelUtil.findObjectFromObjectByID(cm.MainObject, objId);
        return result;
    };
    return CommonModelUtil;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BaseConst = /** @class */ (function () {
    function BaseConst() {
    }
    BaseConst.ID = 'ID';
    BaseConst.Name = 'Name';
    BaseConst.Code = 'Code';
    // 标签
    BaseConst.BeLabel = 'BeLabel';
    return BaseConst;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CmElementConst = /** @class */ (function (_super) {
    __extends(CmElementConst, _super);
    function CmElementConst() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // 字段
    CmElementConst.LabelID = 'LabelID';
    CmElementConst.IsMultiLanguage = 'IsMultiLanguage';
    CmElementConst.DefaultValue = 'DefaultValue';
    CmElementConst.DefaultValueType = 'DefaultValueType';
    CmElementConst.IsRequire = 'IsRequire';
    CmElementConst.Readonly = 'Readonly';
    CmElementConst.IsVirtual = 'IsVirtual';
    // 数据类型
    CmElementConst.MDataType = 'MDataType';
    CmElementConst.Length = 'Length';
    CmElementConst.Precision = 'Precision';
    // 对象类型
    CmElementConst.ObjectType = 'ObjectType';
    CmElementConst.ContainEnumValues = 'ContainEnumValues';
    CmElementConst.EnumIndexType = 'EnumIndexType';
    CmElementConst.ChildAssociations = 'ChildAssociations';
    // 业务字段
    CmElementConst.IsUdt = 'IsUdt';
    CmElementConst.UdtPkgName = 'UdtPkgName';
    CmElementConst.UdtID = 'UdtID';
    CmElementConst.UnifiedDataType = 'UnifiedDataType';
    // static BelongObject = 'BelongObject';
    // 关联
    CmElementConst.KeyCollection = 'KeyCollection';
    CmElementConst.RefElementCollection = 'RefElementCollection';
    CmElementConst.BelongElement = 'BelongElement';
    CmElementConst.RefModelID = 'RefModelID';
    CmElementConst.RefModelName = 'RefModelName';
    CmElementConst.RefModelCode = 'RefModelCode';
    CmElementConst.RefModelPkgName = 'RefModelPkgName';
    // 关联外键
    CmElementConst.TargetElement = 'TargetElement';
    CmElementConst.SourceElement = 'SourceElement';
    CmElementConst.TargetElementDisplay = 'TargetElementDisplay';
    CmElementConst.SourceElementDisplay = 'SourceElementDisplay';
    CmElementConst.RefdataModelName = 'RefdataModelName';
    // 编号规则
    CmElementConst.BillCodeConfig = 'BillCodeConfig';
    // 动态属性集合
    CmElementConst.DynamicPropSetInfo = 'DynamicPropSetInfo';
    // static EnableDynamicProp = 'EnableDynamicProp';
    CmElementConst.DynamicPropRepositoryComp = 'DynamicPropRepositoryComp';
    CmElementConst.DynamicPropSerializerComp = 'DynamicPropSerializerComp';
    return CmElementConst;
}(BaseConst));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CmConst = /** @class */ (function (_super) {
    __extends(CmConst, _super);
    function CmConst() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // 模型
    CmConst.MainObject = 'MainObject';
    CmConst.ComponentAssemblyName = 'ComponentAssemblyName';
    CmConst.AssemblyName = 'AssemblyName';
    CmConst.GeneratingAssembly = 'GeneratingAssembly';
    // 权限
    CmConst.Authorizations = 'Authorizations';
    // 依赖be
    CmConst.Category = 'Category';
    CmConst.DependentEntityId = 'DependentEntityId';
    CmConst.DependentEntityName = 'DependentEntityName';
    CmConst.DependentEntityPackageName = 'DependentEntityPackageName';
    // 生成时间戳字段
    CmConst.IsUsingTimeStamp = 'IsUsingTimeStamp';
    // 变量
    CmConst.Variables = 'Variables';
    // 操作
    CmConst.Description = 'Description';
    CmConst.RequestElements = 'RequestElements';
    CmConst.IsGenerateComponent = 'IsGenerateComponent';
    return CmConst;
}(BaseConst));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CmObjectConst = /** @class */ (function (_super) {
    __extends(CmObjectConst, _super);
    function CmObjectConst() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // 对象
    // static ParentObject = 'ParentObject';
    CmObjectConst.ContainChildObjects = 'ContainChildObjects';
    CmObjectConst.ContainElements = 'ContainElements';
    CmObjectConst.ParentObjectID = 'ParentObjectID';
    CmObjectConst.ObjectType = 'ObjectType';
    CmObjectConst.IsVirtual = 'IsVirtual';
    // 唯一性约束
    CmObjectConst.ContainConstraints = 'ContainConstraints';
    CmObjectConst.ConstraintMessage = 'ConstraintMessage';
    CmObjectConst.Type = 'Type';
    CmObjectConst.ElementList = 'ElementList';
    // 权限
    CmObjectConst.AuthFieldInfos = 'AuthFieldInfos';
    CmObjectConst.MdRefInfo_ID = 'ID';
    CmObjectConst.MdRefInfo_Name = 'Name';
    // 操作
    CmObjectConst.DtmBeforeSave = 'DtmBeforeSave';
    CmObjectConst.DtmAfterModify = 'DtmAfterModify';
    CmObjectConst.DtmAfterCreate = 'DtmAfterCreate';
    /**
     * 持久化构件
     */
    CmObjectConst.RepositoryComps = 'RepositoryComps';
    return CmObjectConst;
}(BaseConst));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 节点元数据
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 节点元数据
 */
var  /**
 * 节点元数据
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 节点元数据
 */
GspCommonDataType = /** @class */ (function () {
    function GspCommonDataType() {
    }
    return GspCommonDataType;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 字段数据类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
var GSPElementDataType = {
    /**
     * 文本
     */
    String: 'String',
    /**
     * 备注
     */
    Text: 'Text',
    /**
     * 整数
     */
    Integer: 'Integer',
    /**
     * 浮点数
     */
    Decimal: 'Decimal',
    /**
     * 布尔型
     */
    Boolean: 'Boolean',
    /**
     * 日期型
     */
    Date: 'Date',
    /**
     * 日期时间型
     */
    DateTime: 'DateTime',
    /**
     * 二进制
     */
    Binary: 'Binary',
};
/** @type {?} */
var GSPElementDataTypes = [
    {
        index: 0, text: '文本', value: GSPElementDataType.String, canChangeLength: true,
        canChangePresicion: false, length: 36, precision: 0
    },
    {
        index: 1, text: '备注', value: GSPElementDataType.Text, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    },
    {
        index: 2, text: '整数', value: GSPElementDataType.Integer, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    },
    {
        index: 3, text: '浮点数字', value: GSPElementDataType.Decimal, canChangeLength: true,
        canChangePresicion: true, length: 18, precision: 2
    },
    {
        index: 4, text: '布尔型', value: GSPElementDataType.Boolean, canChangeLength: false,
        canChangePresicion: false, length: 1, precision: 0
    },
    {
        index: 5, text: '日期', value: GSPElementDataType.Date, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    },
    {
        index: 6, text: '日期时间', value: GSPElementDataType.DateTime, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    },
    {
        index: 7, text: '二进制', value: GSPElementDataType.Binary, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    }, {
        index: 8, text: '业务字段', value: 'UnifiedDataType', canChangeLength: false,
        canChangePresicion: false, length: 36, precision: 0
    }
];
var GetDataTypesUtil = /** @class */ (function () {
    function GetDataTypesUtil() {
    }
    /**
     * @return {?}
     */
    GetDataTypesUtil.GSPElementDataTypesWithouUdt = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var types = new Array();
        Object.assign(types, GSPElementDataTypes);
        types = types.slice(0, 8);
        return types;
    };
    return GetDataTypesUtil;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 字段类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
var GSPElementObjectType = {
    /**
     * 未设置
     */
    None: 'None',
    /**
     * 关联
     */
    Association: 'Association',
    /**
     * 枚举
     */
    Enum: 'Enum',
    /**
     * 动态属性
     */
    DynamicProp: 'DynamicProp',
};
/** @type {?} */
var GSPElementObjectTypes = [
    { index: 0, text: '无', value: GSPElementObjectType.None },
    { index: 1, text: '关联', value: GSPElementObjectType.Association },
    { index: 2, text: '枚举', value: GSPElementObjectType.Enum },
    { index: 3, text: '动态属性', value: GSPElementObjectType.DynamicProp },
];
var GetElementObjTypesUtil = /** @class */ (function () {
    function GetElementObjTypesUtil() {
    }
    /**
     * @return {?}
     */
    GetElementObjTypesUtil.GSPElementObjectTypesWithouDynamic = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var types = new Array();
        Object.assign(types, GSPElementObjectTypes);
        types = types.slice(0, 3);
        return types;
    };
    return GetElementObjTypesUtil;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 字段定义
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 字段定义
 */
var  /**
 * 字段定义
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 字段定义
 */
GspCommonField = /** @class */ (function () {
    function GspCommonField() {
        /**
         * 字段数据类型(默认string)
         */
        this.MDataType = GSPElementDataType.String;
        /**
         * 字段对象类型(默认None)
         */
        this.ObjectType = GSPElementObjectType.None;
    }
    Object.defineProperty(GspCommonField.prototype, "HasAssociation", {
        // private _childAssociations: GSPAssociation[];
        // get ChildAssociations() { if (!this._childAssociations)
        // { this._childAssociations = Array<GSPAssociation>() } return this._childAssociations; }
        // set ChildAssociations(value: GSPAssociation[]) { this._childAssociations = value; }
        /**
         * 包含的关联列表中是否存在关联
         */
        get: 
        // private _childAssociations: GSPAssociation[];
        // get ChildAssociations() { if (!this._childAssociations)
        // { this._childAssociations = Array<GSPAssociation>() } return this._childAssociations; }
        // set ChildAssociations(value: GSPAssociation[]) { this._childAssociations = value; }
        /**
         * 包含的关联列表中是否存在关联
         * @return {?}
         */
        function () {
            if (this.ChildAssociations) {
                return this.ChildAssociations.length > 0;
            }
            else {
                return false;
            }
        },
        enumerable: true,
        configurable: true
    });
    return GspCommonField;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 外键约束类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {number} */
var ForeignKeyConstraintType = {
    /**
     * 禁用
     */
    Forbid: 0,
    /**
     * 启用
     */
    Permit: 1,
};
ForeignKeyConstraintType[ForeignKeyConstraintType.Forbid] = 'Forbid';
ForeignKeyConstraintType[ForeignKeyConstraintType.Permit] = 'Permit';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 关联关系
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 关联关系
 */
var  /**
 * 关联关系
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 关联关系
 */
CefGSPAssociation = /** @class */ (function () {
    function CefGSPAssociation() {
    }
    return CefGSPAssociation;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 关联外键
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 关联外键
 */
var  /**
 * 关联外键
 */
GSPAssociationKey = /** @class */ (function () {
    function GSPAssociationKey() {
    }
    return GSPAssociationKey;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 删除规则类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {number} */
var GSPDeleteRuleType = {
    /**
     * 拒绝删除
     */
    Refuse: 0,
    /**
     * 允许删除
     */
    Allow: 1,
};
GSPDeleteRuleType[GSPDeleteRuleType.Refuse] = 'Refuse';
GSPDeleteRuleType[GSPDeleteRuleType.Allow] = 'Allow';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 变量字段数据类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
var GSPVmVarDataType = {
    /**
     * 文本
     */
    String: 'String',
    /**
     * 备注
     */
    Text: 'Text',
    /**
     * 整数
     */
    Integer: 'Integer',
    /**
     * 浮点数
     */
    Decimal: 'Decimal',
    /**
     * 布尔型
     */
    Boolean: 'Boolean',
    /**
     * 日期型
     */
    Date: 'Date',
    /**
     * 日期时间型
     */
    DateTime: 'DateTime',
    /**
     * 二进制
     */
    Binary: 'Binary',
};
/** @type {?} */
var GSPVmVarDataTypes = [
    {
        index: 0, text: '文本', value: GSPVmVarDataType.String, canChangeLength: true,
        canChangePresicion: false, length: 36, precision: 0
    },
    {
        index: 1, text: '备注', value: GSPVmVarDataType.Text, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    },
    {
        index: 2, text: '整数', value: GSPVmVarDataType.Integer, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    },
    {
        index: 3, text: '浮点数字', value: GSPVmVarDataType.Decimal, canChangeLength: true,
        canChangePresicion: true, length: 18, precision: 2
    },
    {
        index: 4, text: '布尔型', value: GSPVmVarDataType.Boolean, canChangeLength: false,
        canChangePresicion: false, length: 1, precision: 0
    },
    {
        index: 5, text: '日期', value: GSPVmVarDataType.Date, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    },
    {
        index: 6, text: '日期时间', value: GSPVmVarDataType.DateTime, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    },
    {
        index: 7, text: '二进制', value: GSPVmVarDataType.Binary, canChangeLength: false,
        canChangePresicion: false, length: 0, precision: 0
    }
    // , {
    //     index: 8, text: '业务字段', value: 'UnifiedDataType', canChangeLength: false,
    //     canChangePresicion: false, length: 36, precision: 0
    // }
];
var GetVariableDataTypesUtil = /** @class */ (function () {
    function GetVariableDataTypesUtil() {
    }
    /**
     * @return {?}
     */
    GetVariableDataTypesUtil.GSPElementDataTypesWithouUdt = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var types = new Array();
        Object.assign(types, GSPVmVarDataTypes);
        types = types.slice(0, 8);
        return types;
    };
    return GetVariableDataTypesUtil;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 枚举
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 枚举
 */
var  /**
 * 枚举
 */
GSPEnumValue = /** @class */ (function () {
    function GSPEnumValue() {
    }
    return GSPEnumValue;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 联接方式
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {number} */
var JoinMode = {
    /**
     * 内联接
     */
    InnerJoin: 0,
    /**
     * 外联接
     */
    OuterJoin: 1,
};
JoinMode[JoinMode.InnerJoin] = 'InnerJoin';
JoinMode[JoinMode.OuterJoin] = 'OuterJoin';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 动态属性集合构件
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 动态属性集合构件
 */
var  /**
 * 动态属性集合构件
 */
MdRefInfo = /** @class */ (function () {
    function MdRefInfo() {
    }
    return MdRefInfo;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 动态属性信息
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 动态属性信息
 */
var  /**
 * 动态属性信息
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 动态属性信息
 */
DynamicPropSetInfo = /** @class */ (function () {
    function DynamicPropSetInfo() {
    }
    return DynamicPropSetInfo;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
* BE操作基类.
* BE操作，是基于服务端业务逻辑构件的，每个操作都是其中的一个方法。
* 但BE对用户隐藏了该层关系，即让用户感觉不到后台构件的存在。
* 应该确保BE操作的Code、Name与Method的Code、Name的一致。
* @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * BE操作基类.
 * BE操作，是基于服务端业务逻辑构件的，每个操作都是其中的一个方法。
 * 但BE对用户隐藏了该层关系，即让用户感觉不到后台构件的存在。
 * 应该确保BE操作的Code、Name与Method的Code、Name的一致。
 */
var  /**
 * BE操作基类.
 * BE操作，是基于服务端业务逻辑构件的，每个操作都是其中的一个方法。
 * 但BE对用户隐藏了该层关系，即让用户感觉不到后台构件的存在。
 * 应该确保BE操作的Code、Name与Method的Code、Name的一致。
 */
CommonOperation = /** @class */ (function () {
    function CommonOperation() {
        // /**
        //  * 是否可见
        //  */
        // public IsVisible: boolean;
        // /**
        //  * 操作类型
        //  */
        // public OpType: BEOperationType;
        /**
             * 所属的Node节点
             */
        // public Owner: GSPCommonObject;
        // /**
        //  * 所属扩展模型元数据ID
        //  */
        // public BelongModelID: string;
        // /**
        //  * 是否引用
        //  */
        // public IsRef: boolean;
        /**
         * 是否生成构件
         */
        this.IsGenerateComponent = true;
        // this.ID = Guid.newGuid().toString();
        // this.IsRef = false;
        // this.IsGenerateComponent = true;
    }
    return CommonOperation;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
* 联动规则元数据定义
* @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 联动规则元数据定义
 *
 */
var  /**
* 联动规则元数据定义
* @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 联动规则元数据定义
 *
 */
CommonDetermination = /** @class */ (function (_super) {
    __extends(CommonDetermination, _super);
    function CommonDetermination() {
        return _super.call(this) || this;
    }
    return CommonDetermination;
}(CommonOperation));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
* 触发时机，供Validation、Determination共用。
* @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {number} */
var ExecutingDataStatus = {
    /**
     * 不执行
     */
    None: 0,
    /**
     * 实例创建时执行
     */
    Created: 1,
    /**
     * 实例更新时执行
     */
    Updated: 2,
    /**
     * 实例删除时执行
     */
    Deleted: 4,
};
ExecutingDataStatus[ExecutingDataStatus.None] = 'None';
ExecutingDataStatus[ExecutingDataStatus.Created] = 'Created';
ExecutingDataStatus[ExecutingDataStatus.Updated] = 'Updated';
ExecutingDataStatus[ExecutingDataStatus.Deleted] = 'Deleted';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 字段定义
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 字段定义
 */
var  /**
 * 字段定义
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 字段定义
 */
GSPCommonElement = /** @class */ (function (_super) {
    __extends(GSPCommonElement, _super);
    function GSPCommonElement() {
        var _this = _super.call(this) || this;
        /**
         * 字段对象类型(默认None)
         */
        _this.ObjectType = GSPElementObjectType.None;
        // this.ID = Guid.newGuid();
        _this.Length = 36;
        _this.Precision = 0;
        _this.MDataType = GSPElementDataType.String;
        _this.ObjectType = GSPElementObjectType.None;
        _this.IsMultiLanguage = false;
        _this.IsRequire = false;
        _this.IsRef = false;
        return _this;
    }
    return GSPCommonElement;
}(GspCommonField));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 通用模型，作为业务实体，视图模型和数据模型的总接口，用于规划实体数据结构
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 通用模型，作为业务实体，视图模型和数据模型的总接口，用于规划实体数据结构
 */
var  /**
 * 通用模型，作为业务实体，视图模型和数据模型的总接口，用于规划实体数据结构
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 通用模型，作为业务实体，视图模型和数据模型的总接口，用于规划实体数据结构
 */
GSPCommonModel = /** @class */ (function () {
    function GSPCommonModel() {
    }
    // /**
    //  * 根据elementId查找字段
    //  * @param id
    //  */
    // public findElementById(elementId: string): IGSPCommonElement {
    //     let result: IGSPCommonElement = null;
    //     result = this.findElementFromObjectByID(this.MainObject, elementId);
    //     return result;
    // }
    /**
     * 根据objId查找对象
     * @param objId
     */
    // /**
    //  * 根据elementId查找字段
    //  * @param id
    //  */
    // public findElementById(elementId: string): IGSPCommonElement {
    //     let result: IGSPCommonElement = null;
    //     result = this.findElementFromObjectByID(this.MainObject, elementId);
    //     return result;
    // }
    /**
     * 根据objId查找对象
     * @param {?} objId
     * @return {?}
     */
    GSPCommonModel.prototype.findObjectById = 
    // /**
    //  * 根据elementId查找字段
    //  * @param id
    //  */
    // public findElementById(elementId: string): IGSPCommonElement {
    //     let result: IGSPCommonElement = null;
    //     result = this.findElementFromObjectByID(this.MainObject, elementId);
    //     return result;
    // }
    /**
     * 根据objId查找对象
     * @param {?} objId
     * @return {?}
     */
    function (objId) {
        /** @type {?} */
        var result = null;
        result = this.findObjectFromObjectByID(this.MainObject, objId);
        return result;
    };
    // // public getAllObjectList(): IGSPCommonObject[] {
    // //     throw new Error('Method not implemented.');
    // // }
    // public getAllElementList(): IGSPCommonElement[] {
    //     return this.MainObject.getAllObjectElement();
    // }
    // private findElementFromObjectByID(obj: IGSPCommonObject, id: string): IGSPCommonElement {
    //     let result: IGSPCommonElement = null;
    //     //查找当前节点字段集合
    //     for (let i = 0; i < obj.ContainElements.length; i++) {
    //         if (obj.ContainElements[i].ID == id) {
    //             result = obj.ContainElements[i];
    //             return result;
    //         }
    //     }
    //     //递归查找子节点
    //     if (obj.ContainChildObjects.length > 0) {
    //         for (let i = 0; i < obj.ContainChildObjects.length; i++) {
    //             result = this.findElementFromObjectByID(obj.ContainChildObjects[i], id);
    //             if (result != null) {
    //                 return result;
    //             }
    //         }
    //     }
    //     return result;
    // }
    // private findElementFromElementByID(ele: IGSPCommonElement, eleId: string): IGSPCommonElement {
    //     let result: IGSPCommonElement = null;
    //     if (ele.ID == eleId) {
    //         result = ele;
    //     }
    //     else {
    //         //查找所有关联字段及关联的关联字段……
    //         if (ele.HasAssociation) {
    //             for (let i = 0; i < ele.ChildAssociations.length; i++) {
    //                 let asscociation = ele.ChildAssociations[i];
    //                 for (let j = 0; j < asscociation.RefElementCollection.length; j++) {
    //                     let element = this.findElementFromElementByID(asscociation.RefElementCollection[i], eleId);
    //                     if (element != null) {
    //                         result = element;
    //                         return result;
    //                     }
    //                 }
    //             }
    //         }
    //     }
    //     return result;
    // }
    // // public getAllObjectList(): IGSPCommonObject[] {
    // //     throw new Error('Method not implemented.');
    // // }
    // public getAllElementList(): IGSPCommonElement[] {
    //     return this.MainObject.getAllObjectElement();
    // }
    // private findElementFromObjectByID(obj: IGSPCommonObject, id: string): IGSPCommonElement {
    //     let result: IGSPCommonElement = null;
    //     //查找当前节点字段集合
    //     for (let i = 0; i < obj.ContainElements.length; i++) {
    //         if (obj.ContainElements[i].ID == id) {
    //             result = obj.ContainElements[i];
    //             return result;
    //         }
    //     }
    //     //递归查找子节点
    //     if (obj.ContainChildObjects.length > 0) {
    //         for (let i = 0; i < obj.ContainChildObjects.length; i++) {
    //             result = this.findElementFromObjectByID(obj.ContainChildObjects[i], id);
    //             if (result != null) {
    //                 return result;
    //             }
    //         }
    //     }
    //     return result;
    // }
    // private findElementFromElementByID(ele: IGSPCommonElement, eleId: string): IGSPCommonElement {
    //     let result: IGSPCommonElement = null;
    //     if (ele.ID == eleId) {
    //         result = ele;
    //     }
    //     else {
    //         //查找所有关联字段及关联的关联字段……
    //         if (ele.HasAssociation) {
    //             for (let i = 0; i < ele.ChildAssociations.length; i++) {
    //                 let asscociation = ele.ChildAssociations[i];
    //                 for (let j = 0; j < asscociation.RefElementCollection.length; j++) {
    //                     let element = this.findElementFromElementByID(asscociation.RefElementCollection[i], eleId);
    //                     if (element != null) {
    //                         result = element;
    //                         return result;
    //                     }
    //                 }
    //             }
    //         }
    //     }
    //     return result;
    // }
    /**
     * @private
     * @param {?} obj
     * @param {?} objId
     * @return {?}
     */
    GSPCommonModel.prototype.findObjectFromObjectByID = 
    // // public getAllObjectList(): IGSPCommonObject[] {
    // //     throw new Error('Method not implemented.');
    // // }
    // public getAllElementList(): IGSPCommonElement[] {
    //     return this.MainObject.getAllObjectElement();
    // }
    // private findElementFromObjectByID(obj: IGSPCommonObject, id: string): IGSPCommonElement {
    //     let result: IGSPCommonElement = null;
    //     //查找当前节点字段集合
    //     for (let i = 0; i < obj.ContainElements.length; i++) {
    //         if (obj.ContainElements[i].ID == id) {
    //             result = obj.ContainElements[i];
    //             return result;
    //         }
    //     }
    //     //递归查找子节点
    //     if (obj.ContainChildObjects.length > 0) {
    //         for (let i = 0; i < obj.ContainChildObjects.length; i++) {
    //             result = this.findElementFromObjectByID(obj.ContainChildObjects[i], id);
    //             if (result != null) {
    //                 return result;
    //             }
    //         }
    //     }
    //     return result;
    // }
    // private findElementFromElementByID(ele: IGSPCommonElement, eleId: string): IGSPCommonElement {
    //     let result: IGSPCommonElement = null;
    //     if (ele.ID == eleId) {
    //         result = ele;
    //     }
    //     else {
    //         //查找所有关联字段及关联的关联字段……
    //         if (ele.HasAssociation) {
    //             for (let i = 0; i < ele.ChildAssociations.length; i++) {
    //                 let asscociation = ele.ChildAssociations[i];
    //                 for (let j = 0; j < asscociation.RefElementCollection.length; j++) {
    //                     let element = this.findElementFromElementByID(asscociation.RefElementCollection[i], eleId);
    //                     if (element != null) {
    //                         result = element;
    //                         return result;
    //                     }
    //                 }
    //             }
    //         }
    //     }
    //     return result;
    // }
    /**
     * @private
     * @param {?} obj
     * @param {?} objId
     * @return {?}
     */
    function (obj, objId) {
        /** @type {?} */
        var result = null;
        if (!obj) {
            return null;
        }
        if (obj.ID === objId) {
            return obj;
        }
        if (!obj.ContainChildObjects) {
            return null;
        }
        // 递归查找子节点
        if (obj.ContainChildObjects.length > 0) {
            for (var i = 0; i < obj.ContainChildObjects.length; i++) {
                if (obj.ContainChildObjects[i].ID === objId) {
                    result = obj.ContainChildObjects[i];
                    return result;
                }
                result = this.findObjectFromObjectByID(obj.ContainChildObjects[i], objId);
                if (result != null) {
                    return result;
                }
            }
        }
        return result;
    };
    return GSPCommonModel;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 节点元数据
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 节点元数据
 */
var  /**
 * 节点元数据
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 节点元数据
 */
GSPCommonObject = /** @class */ (function (_super) {
    __extends(GSPCommonObject, _super);
    function GSPCommonObject() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(GSPCommonObject.prototype, "IDElement", {
        /**
         * 当前节点主键字段
         */
        get: /**
         * 当前节点主键字段
         * @return {?}
         */
        function () {
            var _this = this;
            if (this.ColumnGenerateID != null) {
                return this.ContainElements.find((/**
                 * @param {?} ele
                 * @return {?}
                 */
                function (ele) { return ele.ID === _this.ColumnGenerateID.ElementID; }));
            }
        },
        enumerable: true,
        configurable: true
    });
    return GSPCommonObject;
}(GspCommonDataType));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 外键约束关系
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 外键约束关系
 * @abstract
 */
var  /**
 * 外键约束关系
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 外键约束关系
 * @abstract
 */
Constraint = /** @class */ (function () {
    function Constraint() {
    }
    return Constraint;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 外键约束关系
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 外键约束关系
 * @template FKeyResourceType, RefResourceType
 */
var  /**
 * 外键约束关系
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 外键约束关系
 * @template FKeyResourceType, RefResourceType
 */
ForeignKeyConstraint = /** @class */ (function (_super) {
    __extends(ForeignKeyConstraint, _super);
    function ForeignKeyConstraint() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ForeignKeyConstraint;
}(Constraint));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 关联外键约束
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 关联外键约束
 */
var  /**
 * 关联外键约束
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 关联外键约束
 */
RelationForeignKeyConstraint = /** @class */ (function (_super) {
    __extends(RelationForeignKeyConstraint, _super);
    function RelationForeignKeyConstraint() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return RelationForeignKeyConstraint;
}(ForeignKeyConstraint));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 约束信息
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 约束信息
 * @abstract
 */
var  /**
 * 约束信息
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 约束信息
 * @abstract
 */
Resource = /** @class */ (function () {
    function Resource() {
    }
    return Resource;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 关联约束信息
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 关联约束信息
 */
var  /**
 * 关联约束信息
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 关联约束信息
 */
RelationResource = /** @class */ (function (_super) {
    __extends(RelationResource, _super);
    function RelationResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return RelationResource;
}(Resource));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 代码生成类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {number} */
var GSPBillCodeGenerateType = {
    /**
     * 未设置
     */
    none: 0,
    /**
     * 系统生成
     */
    genersoft: 1,
    /**
     * 手工生成
     */
    hand: 2,
    /**
     * 自增
     */
    increment: 3,
};
GSPBillCodeGenerateType[GSPBillCodeGenerateType.none] = 'none';
GSPBillCodeGenerateType[GSPBillCodeGenerateType.genersoft] = 'genersoft';
GSPBillCodeGenerateType[GSPBillCodeGenerateType.hand] = 'hand';
GSPBillCodeGenerateType[GSPBillCodeGenerateType.increment] = 'increment';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 代码生成规则执行时机
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
var GSPBillCodeGenerateOccasion = {
    /**
     * 系统定义
     */
    SystemProcess: 'SystemProcess',
    /**
     * 创建时
     */
    CreatingTime: 'CreatingTime',
    /**
     * 保存时
     */
    SavingTime: 'SavingTime',
    /**
     * 创建和保存时
     */
    BothCreatingAndSaving: 'BothCreatingAndSaving',
};
/** @type {?} */
var GSPBillCodeGenerateOccasions = [
    { text: '系统处理', value: GSPBillCodeGenerateOccasion.SystemProcess },
    { text: '创建时', value: GSPBillCodeGenerateOccasion.CreatingTime },
    { text: '保存时', value: GSPBillCodeGenerateOccasion.SavingTime },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 编码规则配置
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 编码规则配置
 */
var  /**
 * 编码规则配置
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 编码规则配置
 */
ElementCodeRuleConfig = /** @class */ (function () {
    function ElementCodeRuleConfig() {
        this.CanBillCode = false;
        this.BillCodeID = '';
        this.BillCodeName = '';
        this.CodeGenerateType = GSPBillCodeGenerateType.none;
        this.CodeGenerateOccasion = GSPBillCodeGenerateOccasion.SystemProcess;
    }
    return ElementCodeRuleConfig;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 字段默认值类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {number} */
var ElementDefaultVauleType = {
    /**
     * 值类型
     */
    Value: 0,
    /**
     * 表达式类型
     */
    Expression: 1,
};
ElementDefaultVauleType[ElementDefaultVauleType.Value] = 'Value';
ElementDefaultVauleType[ElementDefaultVauleType.Expression] = 'Expression';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 字段分级码生成时机
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
var ElementPathGenerateType = {
    /**
     * 创建时
     */
    CreateNode: 'CreateNode',
    /**
     * 保存时
     */
    SaveNode: 'SaveNode',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 关联关系
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 关联关系
 */
var  /**
 * 关联关系
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 关联关系
 */
GSPAssociation = /** @class */ (function (_super) {
    __extends(GSPAssociation, _super);
    function GSPAssociation() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return GSPAssociation;
}(CefGSPAssociation));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 约束类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {number} */
var ConstraintType = {
    /**
     * 唯一约束
     */
    Unique: 0,
};
ConstraintType[ConstraintType.Unique] = 'Unique';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 字段生成规则
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 字段生成规则
 */
var  /**
 * 字段生成规则
 */
GSPColumnGenerate = /** @class */ (function () {
    function GSPColumnGenerate() {
    }
    return GSPColumnGenerate;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 节点类型
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {number} */
var GSPCommonObjectType = {
    /**
     * 主节点
     */
    MainObject: 0,
    /**
     * 非主节点
     */
    ChildObject: 1,
};
GSPCommonObjectType[GSPCommonObjectType.MainObject] = 'MainObject';
GSPCommonObjectType[GSPCommonObjectType.ChildObject] = 'ChildObject';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 分级信息
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 分级信息
 */
var  /**
 * 分级信息
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 分级信息
 */
GSPHirarchyInfo = /** @class */ (function () {
    function GSPHirarchyInfo() {
    }
    return GSPHirarchyInfo;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 唯一约束
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 唯一约束
 */
var  /**
 * 唯一约束
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 唯一约束
 */
GSPUniqueConstraint = /** @class */ (function () {
    function GSPUniqueConstraint() {
    }
    return GSPUniqueConstraint;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 变量来源
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/** @enum {string} */
var VariableSourceType = {
    /**
     * BE
     */
    BE: 'BE',
    /**
     * VO
     */
    VO: 'VO',
    /**
     * 上下文
     */
    Context: 'Context',
};
/** @type {?} */
var VariableSourceTypes = [
    { index: 0, text: 'BE', value: VariableSourceType.BE },
    { index: 1, text: 'VO', value: VariableSourceType.VO },
    { index: 2, text: 'Context', value: VariableSourceType.Context }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 变量定义
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 变量定义
 */
var  /**
 * 变量定义
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 变量定义
 */
CommonVariable = /** @class */ (function (_super) {
    __extends(CommonVariable, _super);
    function CommonVariable() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return CommonVariable;
}(GspCommonField));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 变量实体定义
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 变量实体定义
 */
var  /**
 * 变量实体定义
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 变量实体定义
 */
CommonVariableEntity = /** @class */ (function (_super) {
    __extends(CommonVariableEntity, _super);
    function CommonVariableEntity() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return CommonVariableEntity;
}(GspCommonDataType));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var VersionControlInfo = /** @class */ (function () {
    function VersionControlInfo() {
    }
    return VersionControlInfo;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var EnumIndexType = {
    /**
     * 整型
     */
    Integer: 0,
    /**
     * 字符串
     */
    String: 1,
};
EnumIndexType[EnumIndexType.Integer] = 'Integer';
EnumIndexType[EnumIndexType.String] = 'String';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { CommonModelUtil, BaseConst, CmElementConst, CmConst, CmObjectConst, GSPCommonElement, GSPCommonModel, GSPCommonObject, Constraint, ForeignKeyConstraint, RelationForeignKeyConstraint, RelationResource, Resource, ElementCodeRuleConfig, ElementDefaultVauleType, ElementPathGenerateType, GSPBillCodeGenerateOccasion, GSPBillCodeGenerateOccasions, GSPBillCodeGenerateType, GSPAssociation, ConstraintType, GSPColumnGenerate, GSPCommonObjectType, GSPHirarchyInfo, GSPUniqueConstraint, VariableSourceType, VariableSourceTypes, CommonVariable, CommonVariableEntity, VersionControlInfo, GspCommonDataType, GspCommonField, ForeignKeyConstraintType, CefGSPAssociation, GSPAssociationKey, GSPDeleteRuleType, GSPElementDataType, GSPElementDataTypes, GetDataTypesUtil, GSPVmVarDataType, GSPVmVarDataTypes, GetVariableDataTypesUtil, GetElementObjTypesUtil, GSPElementObjectType, GSPElementObjectTypes, GSPEnumValue, EnumIndexType, JoinMode, MdRefInfo, DynamicPropSetInfo, CommonDetermination, CommonOperation, ExecutingDataStatus };
