// @farris/ide-devkit/esm5/lib/services/http/http.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { CacheService } from '../cache/services/cache.service';
import { map } from 'rxjs/operators';
var HttpService = /** @class */ (function () {
    function HttpService(http, cache) {
        this.http = http;
        this.cache = cache;
    }
    /**
     * @param {?} url
     * @return {?}
     */
    HttpService.prototype.get = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return this.request('Get', url, this.setHeader({}));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    HttpService.prototype.post = /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        return this.request('Post', url, this.setHeader({ body: body }));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    HttpService.prototype.put = /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        return this.request('Put', url, this.setHeader({ body: body }));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    HttpService.prototype.delete = /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        return this.request('Delete', url, this.setHeader({ body: body }));
    };
    /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    HttpService.prototype.patch = /**
     * @param {?} url
     * @param {?} body
     * @return {?}
     */
    function (url, body) {
        return this.request('Patch', url, this.setHeader({ body: body }));
    };
    /**
     * @param {?} options
     * @return {?}
     */
    HttpService.prototype.setHeader = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        options.headers = {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*'
        };
        if (this.cache.exists('sessionId') && this.cache.get('sessionId')) {
            options.headers['SessionId'] = this.cache.get('sessionId');
        }
        return options;
    };
    /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    HttpService.prototype.request = /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @return {?}
     */
    function (method, url, options) {
        if (options.body) {
            if (typeof options.body !== 'string') {
                options.body = JSON.stringify(options.body);
            }
        }
        return this.http.request(method, url, options)
            .pipe(map(function (data) {
            console.log(data);
            return data;
        }));
    };
    HttpService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    HttpService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: CacheService }
    ]; };
    return HttpService;
}());
export { HttpService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    HttpService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    HttpService.prototype.cache;
}
