// @farris/devkit/esm2015/lib/command/command_handler_factory.js
/**
 * 命令处理器工厂相关定义
 * @author Witt<jiwt@inspur.com>
 */
import { Injectable } from '@angular/core';
import { FrameContext } from '../frame/index';
import { CommandHandlerRegistry } from './command_handler_registry';
import { CommandHandlerExtenderRegistry } from './command_handler_extender_registry';
import { VariableParseService } from '../variable/variable_parse_service';
/**
 * 命令处理器工厂
 */
class CommandHandlerFactory {
    /**
     * 构造函数
     * @param handlerRegistry  命令处理注册器
     * @param extenderRegistry 命令处理扩展注册器
     */
    constructor(handlerRegistry, extenderRegistry, frameContext, variableParseService) {
        this.handlerRegistry = handlerRegistry;
        this.extenderRegistry = extenderRegistry;
        this.frameContext = frameContext;
        this.variableParseService = variableParseService;
    }
    /**
     * 创建命令处理器
     * @param   commandName 命令名称
     * @returns 对应的命令处理器实例
     */
    create(commandName) {
        const rawHandler = this.handlerRegistry.get(commandName);
        rawHandler.init(this.frameContext, this.variableParseService);
        const extenders = this.extenderRegistry.get(commandName);
        // 遍历extenders，依次对handler进行扩展
        return extenders.reduce((handler, extender) => {
            return extender.extend(handler);
        }, rawHandler);
    }
    dispose() {
        this.handlerRegistry.dispose();
        this.extenderRegistry.dispose();
        this.frameContext = null;
        this.variableParseService = null;
    }
}
CommandHandlerFactory.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CommandHandlerFactory.ctorParameters = () => [
    { type: CommandHandlerRegistry },
    { type: CommandHandlerExtenderRegistry },
    { type: FrameContext },
    { type: VariableParseService }
];
export { CommandHandlerFactory };
