// @gsp-sys/sysmgr-ui/bundles/gsp-sys-sysmgr-ui.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@farris/ui-layout'), require('@progress/kendo-angular-treeview'), require('@progress/kendo-angular-grid'), require('@progress/kendo-angular-layout'), require('@angular/common'), require('@farris/ui-modal'), require('@farris/ui-messager'), require('@farris/ui-notify'), require('@farris/ui-treetable'), require('@angular/common/http'), require('@angular/forms'), require('@angular/core'), require('@farris/ui-lookup'), require('@gsp-sys/sysmgr-common'), require('rxjs'), require('rxjs/operators'), require('@farris/ui-datagrid'), require('@ecp-caf/caf-common'), require('ngx-cookie-service')) :
    typeof define === 'function' && define.amd ? define('@gsp-sys/sysmgr-ui', ['exports', '@farris/ui-layout', '@progress/kendo-angular-treeview', '@progress/kendo-angular-grid', '@progress/kendo-angular-layout', '@angular/common', '@farris/ui-modal', '@farris/ui-messager', '@farris/ui-notify', '@farris/ui-treetable', '@angular/common/http', '@angular/forms', '@angular/core', '@farris/ui-lookup', '@gsp-sys/sysmgr-common', 'rxjs', 'rxjs/operators', '@farris/ui-datagrid', '@ecp-caf/caf-common', 'ngx-cookie-service'], factory) :
    (factory((global['gsp-sys'] = global['gsp-sys'] || {}, global['gsp-sys']['sysmgr-ui'] = {}),global.uiLayout,global.kendoAngularTreeview,global.kendoAngularGrid,global.kendoAngularLayout,global.ng.common,global.uiModal,global.uiMessager,global.uiNotify,global.uiTreetable,global.ng.common.http,global.ng.forms,global.ng.core,global.uiLookup,global.i2,global.rxjs,global.rxjs.operators,global.uiDatagrid,global.cafCommon,global.ngxCookieService));
}(this, (function (exports,uiLayout,kendoAngularTreeview,kendoAngularGrid,kendoAngularLayout,common,uiModal,uiMessager,uiNotify,uiTreetable,i1,forms,i0,uiLookup,i2,rxjs,operators,uiDatagrid,cafCommon,ngxCookieService) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SysFrontUtil = /** @class */ (function () {
        function SysFrontUtil() {
        }
        Object.defineProperty(SysFrontUtil.prototype, "operationApi", {
            /** 功能操作 */
            get: /**
             * 功能操作
             * @return {?}
             */ function () {
                return '/api/runtime/sys/v1.0/funcOperations';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SysFrontUtil.prototype, "bizObjectApi", {
            /** 业务对象 */
            get: /**
             * 业务对象
             * @return {?}
             */ function () {
                return '/api/runtime/sys/v1.0/business-objects';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(SysFrontUtil.prototype, "httpHeader", {
            /**
             * Http头信息
             */
            get: /**
             * Http头信息
             * @return {?}
             */ function () {
                /** @type {?} */
                var options = {
                    headers: {
                        'Content-Type': 'application/json',
                        'Access-Control-Allow-Origin': '*'
                        //                SessionId: this.sessionId
                    },
                    responseType: 'json'
                };
                return options;
            },
            enumerable: true,
            configurable: true
        });
        return SysFrontUtil;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SysFuncOperationService = /** @class */ (function () {
        function SysFuncOperationService(sysUtil, rest) {
            this.sysUtil = sysUtil;
            this.rest = rest;
        }
        /**
         * 获取功能操作列表
         */
        /**
         * 获取功能操作列表
         * @return {?}
         */
        SysFuncOperationService.prototype.getFuncOperations = /**
         * 获取功能操作列表
         * @return {?}
         */
            function () {
                /** @type {?} */
                var url = this.sysUtil.operationApi + '/all';
                return this.rest.get(url, this.sysUtil.httpHeader);
            };
        SysFuncOperationService.decorators = [
            { type: i0.Injectable }
        ];
        /** @nocollapse */
        SysFuncOperationService.ctorParameters = function () {
            return [
                { type: SysFrontUtil },
                { type: i1.HttpClient }
            ];
        };
        return SysFuncOperationService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 系统管理：功能操作左树有、右列表展示
     */
    var SysFuncOperationComponent = /** @class */ (function () {
        function SysFuncOperationComponent(funcOpService) {
            this.funcOpService = funcOpService;
            /**
             * 是否只显示分组
             */
            this.onlyShowGroup = false;
            /**
             * 确认事件
             */
            this.afterConfirm = new i0.EventEmitter();
            /**
             * 关闭事件
             */
            this.afterCancle = new i0.EventEmitter();
            /**
             * 当前选中数据
             */
            this.curSelectedKeys = [];
            this.title = '功能操作帮助';
        }
        /**
         * @return {?}
         */
        SysFuncOperationComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () { };
        /**
         * @return {?}
         */
        SysFuncOperationComponent.prototype.loadData = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.selectableSettings = {
                    checkboxOnly: true,
                    // 默认单选
                    mode: !this.gridSelectMode ? 'single' : this.gridSelectMode,
                    enabled: true
                };
                this.funcOpService.getFuncOperations().subscribe(( /**
                 * @param {?} res
                 * @return {?}
                 */function (res) {
                    _this.data = [];
                    /** @type {?} */
                    var funcList = res;
                    if (_this.onlyShowGroup) {
                        _this.data = funcList.filter(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) { return item.type == 0; }));
                    }
                    else {
                        _this.data = funcList;
                    }
                    // funcList.forEach(item => {
                    //     this.data.push(item);
                    //     // 如果有明细节点
                    //     if (this.onlyShowGroup === false && item['children'] !== null && item['children'].length > 0) {
                    //         item['children'].forEach(child => this.data.push(child));
                    //     }
                    // });
                    _this.gridData = [];
                    // 复制一份
                    _this.gridData = _this.data.concat();
                }), ( /**
                 * @param {?} error
                 * @return {?}
                 */function (error) {
                    console.log('获取功能操作对象列表失败');
                }));
            };
        /**
         * 搜索
         * @param filterValue 查询条件
         */
        /**
         * 搜索
         * @param {?} filterValue 查询条件
         * @return {?}
         */
        SysFuncOperationComponent.prototype.searchData = /**
         * 搜索
         * @param {?} filterValue 查询条件
         * @return {?}
         */
            function (filterValue) {
                var _this = this;
                // 测试
                if (filterValue === null || filterValue === '') {
                    // 全量的数据
                    this.gridData = this.data.concat();
                    return;
                }
                /** @type {?} */
                var filter = filterValue.toLowerCase();
                // 搜索到上级时
                this.gridData = this.data.filter(( /**
                 * @param {?} item
                 * @return {?}
                 */function (item) {
                    return item['code'].toLowerCase().includes(filter) ||
                        item['name'].toLowerCase().includes(filter) ||
                        _this.getFuncGroup(item).toLowerCase().includes(filter);
                }));
            };
        /**
         * 业务对象变化后
         * @param arg 业务对象参数
         */
        /**
         * 业务对象变化后
         * @param {?} arg 业务对象参数
         * @return {?}
         */
        SysFuncOperationComponent.prototype.bizObjectChanged = /**
         * 业务对象变化后
         * @param {?} arg 业务对象参数
         * @return {?}
         */
            function (arg) {
                if (!this.data || this.data.length === 0) {
                    return;
                }
                /** @type {?} */
                var busId = arg;
                // 根据业务对象过滤
                this.gridData = this.data.filter(( /**
                 * @param {?} item
                 * @return {?}
                 */function (item) {
                    return item['businessId'] === busId;
                }));
            };
        /**
         * 获取功能分组
         */
        /**
         * 获取功能分组
         * @param {?} funcOpObj
         * @return {?}
         */
        SysFuncOperationComponent.prototype.getFuncGroup = /**
         * 获取功能分组
         * @param {?} funcOpObj
         * @return {?}
         */
            function (funcOpObj) {
                // 获取父节点
                /** @type {?} */
                var parentId = funcOpObj['parentId'];
                if (parentId !== null && parentId !== undefined && parentId !== '') {
                    /** @type {?} */
                    var parentObj = this.data.filter(( /**
                     * @param {?} item
                     * @return {?}
                     */function (item) { return item['id'] === parentId; }));
                    if (parentObj !== null && parentObj.length > 0) {
                        return parentObj[0]['name'];
                    }
                }
                else {
                    return '无';
                }
            };
        /**
         * 获取选中数据
         */
        /**
         * 获取选中数据
         * @return {?}
         */
        SysFuncOperationComponent.prototype.getSelectedObject = /**
         * 获取选中数据
         * @return {?}
         */
            function () {
                var _this = this;
                /** @type {?} */
                var result = this.data.filter(( /**
                 * @param {?} item
                 * @return {?}
                 */function (item) {
                    return _this.curSelectedKeys.findIndex(( /**
                     * @param {?} select
                     * @return {?}
                     */function (select) { return select === item['id']; })) >= 0;
                }));
                return result;
            };
        /**
         * 确认
         */
        /**
         * 确认
         * @return {?}
         */
        SysFuncOperationComponent.prototype.confirm = /**
         * 确认
         * @return {?}
         */
            function () {
                /** @type {?} */
                var result = this.getSelectedObject();
                this.afterConfirm.next(result);
                // 取消-目的是在这里让模态框消失
                this.cancle();
            };
        /**
         * 取消
         */
        /**
         * 取消
         * @return {?}
         */
        SysFuncOperationComponent.prototype.cancle = /**
         * 取消
         * @return {?}
         */
            function () {
                this.afterCancle.next();
            };
        SysFuncOperationComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'app-sys-func-operation',
                        template: "<div class=\"d-flex flex-row\" style=\"\r\nwidth: 800px; \r\nborder:2px;\r\nposition:absolute;\r\ntop:0px;\r\nbottom: 0px;\r\nright:0px;\r\nleft:0px;\r\noverflow: hidden;\r\nheight: 100%;\">\r\n  <div class=\"flex-fill\">\r\n    <kendo-splitter style=\"height:100%\" orientation=\"horizontal\">\r\n      <kendo-splitter-pane size=\"40%\" min=\"20%\" max=\"50%\">\r\n        <div class=\"d-flex flex-column flex-fill\" style=\"height:100%\">\r\n          <div class=\"p-2\" style=\"height: 35px;\">\r\n            <h5>\u4E1A\u52A1\u5BF9\u8C61</h5>\r\n            <hr class=\"m-0 p-0\" />\r\n          </div>\r\n          <div class=\"flex-fill\" style=\"overflow: auto;\">\r\n            <app-biz-object [bizObjectId]=\"bizObjectId\" (bizObjectChanged)=\"bizObjectChanged($event)\"></app-biz-object>\r\n          </div>\r\n        </div>\r\n      </kendo-splitter-pane>\r\n      <kendo-splitter-pane>\r\n        <div class=\"d-flex flex-column flex-fill\" style=\"height:100%\">\r\n          <div class=\"p-2\" style=\"height: 35px;\">\r\n            <h5>\u529F\u80FD\u64CD\u4F5C</h5>\r\n            <hr class=\"m-0 p-0\" />\r\n          </div>\r\n          <div class=\"clearfix mt-1 mb-1 ml-1\">\r\n            <input #filter placeholder=\"\u8F93\u5165\u7F16\u53F7\u6216\u540D\u79F0\u67E5\u627E\" type=\"text\" />\r\n            <button class=\"btn btn-sm btn-default\" (click)=\"searchData(filter.value)\">\u67E5\u627E</button>\r\n          </div>\r\n          <div class=\"d-flex flex-fill\" style=\"position:relative;\">\r\n            <kendo-grid [class]=\"'flex-fill'\" [selectable]=\"selectableSettings\" [kendoGridBinding]=\"gridData\"\r\n              [kendoGridSelectBy]=\"'id'\" [selectedKeys]=\"curSelectedKeys\">\r\n              <kendo-grid-checkbox-column width=\"40\"></kendo-grid-checkbox-column>\r\n              <kendo-grid-column field=\"code\" title=\"\u7F16\u53F7\"></kendo-grid-column>\r\n              <kendo-grid-column field=\"name\" title=\"\u540D\u79F0\"></kendo-grid-column>\r\n              <kendo-grid-column *ngIf=\"!onlyShowGroup\" title=\"\u529F\u80FD\u5206\u7EC4\">\r\n                <ng-template kendoGridCellTemplate let-dataItem>\r\n                  {{getFuncGroup(dataItem)}}\r\n                </ng-template>\r\n              </kendo-grid-column>\r\n            </kendo-grid>\r\n          </div>\r\n        </div>\r\n      </kendo-splitter-pane>\r\n    </kendo-splitter>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #btnOkAndCancle>\r\n  <button class=\"btn btn-sm btn-primary\" (click)=\"confirm()\">\u786E\u8BA4</button>\r\n  <button class=\"btn btn-sm btn-default\" (click)=\"cancle()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        SysFuncOperationComponent.ctorParameters = function () {
            return [
                { type: SysFuncOperationService }
            ];
        };
        SysFuncOperationComponent.propDecorators = {
            bizObjectId: [{ type: i0.Input }],
            gridSelectMode: [{ type: i0.Input }],
            sessionId: [{ type: i0.Input }],
            onlyShowGroup: [{ type: i0.Input }],
            isGroupCanSelected: [{ type: i0.Input }],
            afterConfirm: [{ type: i0.Output }],
            afterCancle: [{ type: i0.Output }],
            btnOkAndCancleRef: [{ type: i0.ViewChild, args: ['btnOkAndCancle',] }]
        };
        return SysFuncOperationComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BizObjectService = /** @class */ (function () {
        function BizObjectService(rest, sysUtil) {
            this.rest = rest;
            this.sysUtil = sysUtil;
        }
        /**
         * 获取业务对象列表
         */
        /**
         * 获取业务对象列表
         * @return {?}
         */
        BizObjectService.prototype.getBizObjects = /**
         * 获取业务对象列表
         * @return {?}
         */
            function () {
                /** @type {?} */
                var url = this.sysUtil.bizObjectApi;
                return this.rest.get(url, this.sysUtil.httpHeader);
            };
        BizObjectService.decorators = [
            { type: i0.Injectable }
        ];
        /** @nocollapse */
        BizObjectService.ctorParameters = function () {
            return [
                { type: i1.HttpClient },
                { type: SysFrontUtil }
            ];
        };
        return BizObjectService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var NAVI_LOCALE_EN_US = {
        rootNodeCode: 'ALL',
        rootNodeTitle: 'all data',
        codeField: 'Code',
        nameField: 'Name',
        allColumnsTitle: 'All columns',
        searchNullData: 'No data to find!'
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var HELP_LOCALE_EN_US = {
        userTitle: 'Users',
        positionTitle: 'Positions',
        roleTitle: 'Roles',
        userOrgTitle: 'Organizations',
        userGroupTitle: 'User Groups',
        positionGroupTitle: 'Position Groups',
        roleGroupTitle: 'Role Groups',
        codeField: 'Code',
        nameField: 'Name',
        orderField: 'Order',
        orgName: 'Org Info',
        orgNamePathField: 'Company Info',
        operationTitle: 'Function Operation',
        busTitle: 'Business Object',
        opGroupField: 'Operation Group'
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var EN_US = {
        navi: NAVI_LOCALE_EN_US,
        help: HELP_LOCALE_EN_US
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var NAVI_LOCALE_ZHCHS = {
        rootNodeCode: 'ALL',
        rootNodeTitle: '所有',
        codeField: '编号',
        nameField: '名称',
        allColumnsTitle: '所有列',
        searchNullData: '没有要查找的数据！'
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var HELP_LOCALE_ZHCHS = {
        userTitle: '用户',
        positionTitle: '岗位',
        roleTitle: '功能组',
        userOrgTitle: '组织',
        userGroupTitle: '用户分组',
        positionGroupTitle: '岗位分组',
        roleGroupTitle: '功能组分组',
        codeField: '编号',
        nameField: '名称',
        orderField: '排序号',
        orgName: '隶属组织',
        orgNamePathField: '隶属单位',
        operationTitle: '功能操作',
        busTitle: '业务对象',
        opGroupField: '功能分组'
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var ZH_CHS = {
        navi: NAVI_LOCALE_ZHCHS,
        help: HELP_LOCALE_ZHCHS
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var NAVI_LOCALE_ZHCHT = {
        rootNodeCode: 'ALL',
        rootNodeTitle: '所有',
        codeField: '編號',
        nameField: '名稱',
        allColumnsTitle: '所有列',
        searchNullData: '沒有要查找的數據！'
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var HELP_LOCALE_ZHCHT = {
        userTitle: '用戶',
        positionTitle: '崗位',
        roleTitle: '功能組',
        userOrgTitle: '組織',
        userGroupTitle: '用戶分組',
        positionGroupTitle: '崗位分組',
        roleGroupTitle: '功能組分組',
        codeField: '編號',
        nameField: '名稱',
        orderField: '排序號',
        orgName: '隸屬組織',
        orgNamePathField: '隸屬單位',
        operationTitle: '功能操作',
        busTitle: '業務對象',
        opGroupField: '功能分組'
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var ZH_CHT = {
        navi: NAVI_LOCALE_ZHCHT,
        help: HELP_LOCALE_ZHCHT
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var SYS_LOCALES = {
        'zh-CHS': ZH_CHS,
        'zh-cn': ZH_CHS,
        en: EN_US,
        'zh-CHT': ZH_CHT
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SysLocaleService = /** @class */ (function () {
        function SysLocaleService(lang) {
            this.lang = lang;
            /** @type {?} */
            var localeId = 'zh-CHS';
            if (lang) {
                localeId = lang.getLanguageCode();
            }
            this.langData = SYS_LOCALES[localeId];
            if (!this.langData) {
                this.langData = SYS_LOCALES['zh-CHS'];
            }
            console.log(localeId);
            console.log(this.langData);
        }
        /**
         * @param {?} comp
         * @return {?}
         */
        SysLocaleService.prototype.getComponentOpt = /**
         * @param {?} comp
         * @return {?}
         */
            function (comp) {
                return this.langData[comp];
            };
        SysLocaleService.decorators = [
            { type: i0.Injectable }
        ];
        /** @nocollapse */
        SysLocaleService.ctorParameters = function () {
            return [
                { type: cafCommon.LanguageService }
            ];
        };
        return SysLocaleService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 系统管理：功能操作左树有、右列表展示
     */
    var SysFarrisFuncOperationComponent = /** @class */ (function () {
        function SysFarrisFuncOperationComponent(funcOpService, bizObjectService, localService) {
            var _this = this;
            this.funcOpService = funcOpService;
            this.bizObjectService = bizObjectService;
            this.localService = localService;
            this.multiSelect = false;
            /**
             * 是否只显示分组
             */
            this.onlyShowGroup = false;
            /**
             * 确认事件
             */
            this.afterConfirm = new i0.EventEmitter();
            /**
             * 关闭事件
             */
            this.afterCancle = new i0.EventEmitter();
            /**
             * 当前选中数据
             */
            this.curSelected = [];
            this.curSelectedKeys = [];
            this.gridColumns = [];
            this.opSearchTxt = null;
            /**
             * 格式化
             */
            this.funcGroupFormatter = ( /**
             * @param {?} val
             * @param {?} data
             * @return {?}
             */function (val, data) {
                return _this.getFuncGroup(data);
            });
        }
        /**
         * 监控拖拽事件，重绘组织TreeTable
         * @param event 事件参数
         */
        /**
         * 监控拖拽事件，重绘组织TreeTable
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.winResize = /**
         * 监控拖拽事件，重绘组织TreeTable
         * @return {?}
         */
            function () {
                this.reSizeTreeTable();
            };
        /** 重绘Farris-TreeTable的大小 */
        /**
         * 重绘Farris-TreeTable的大小
         * @param {?=} height
         * @param {?=} width
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.reSizeTreeTable = /**
         * 重绘Farris-TreeTable的大小
         * @param {?=} height
         * @param {?=} width
         * @return {?}
         */
            function (height, width) {
                if (height === void 0) {
                    height = 0;
                }
                if (width === void 0) {
                    width = 0;
                }
                if (height === 0) {
                    height = this.farrisTreeParent.nativeElement.clientHeight;
                }
                if (width === 0) {
                    width = this.farrisTreeParent.nativeElement.clientWidth;
                }
                this.farrisTt.resize({ height: height, width: width });
            };
        /**
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.operationGrid.restService = null;
                this.operationGrid.virtualized = true;
                this.operationGrid.fitColumns = true;
                this.operationGrid.fit = true;
                if (this.localService) {
                    this.i18nRes = this.localService.getComponentOpt('help');
                }
                else {
                    this.i18nRes = HELP_LOCALE_ZHCHS;
                }
                // 只显示分组，功能菜单使用
                if (this.onlyShowGroup) {
                    this.gridColumns = [
                        { field: 'code', title: this.i18nRes.codeField, width: 120 },
                        { field: 'name', title: this.i18nRes.nameField, width: 150 }
                    ];
                    // 私有方法，更新列
                    // this.operationGrid['columnsChanged']();
                }
                else {
                    this.gridColumns = [
                        { field: 'code', title: this.i18nRes.codeField, width: 120 },
                        { field: 'name', title: this.i18nRes.nameField, width: 150 },
                        { field: 'parentId', title: this.i18nRes.opGroupField, formatter: this.funcGroupFormatter }
                    ];
                }
                this.treeColumns = [
                    { field: 'name', title: this.i18nRes.busTitle, width: 100 }
                ];
                // 多选
                if (this.gridSelectMode === 'multiple') {
                    this.multiSelect = true;
                }
                this.loadData();
            };
        /** 页面渲染完成后 */
        /**
         * 页面渲染完成后
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.ngAfterContentInit = /**
         * 页面渲染完成后
         * @return {?}
         */
            function () {
                this.reSizeTreeTable();
            };
        /**
         * 加载数据,TODO 加载数据
         */
        /**
         * 加载数据,TODO 加载数据
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.loadData = /**
         * 加载数据,TODO 加载数据
         * @return {?}
         */
            function () {
                var _this = this;
                // 加载业务对象数据
                this.bizObjectService.getBizObjects().subscribe(( /**
                 * @param {?} result
                 * @return {?}
                 */function (result) {
                    _this.treeDataSrc = result;
                    _this.treeData = _this.makeTree(result);
                }));
                // 加载功能操作数据
                this.funcOpService.getFuncOperations().subscribe(( /**
                 * @param {?} lists
                 * @return {?}
                 */function (lists) {
                    // 只显示分组
                    if (_this.onlyShowGroup) {
                        _this.gridDataSrc = lists.filter(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) { return item.type == 0; }));
                    }
                    else {
                        _this.gridDataSrc = lists;
                    }
                    // 一定要延迟一段时间
                    _this.freshGridAndSelect(_this.gridDataSrc.concat());
                }), ( /**
                 * @param {?} error
                 * @return {?}
                 */function (error) {
                    console.log('获取功能操作对象列表失败');
                }));
            };
        /**
         * 刷新列表显示
         */
        /**
         * 刷新列表显示
         * @private
         * @param {?} list
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.freshGridAndSelect = /**
         * 刷新列表显示
         * @private
         * @param {?} list
         * @return {?}
         */
            function (list) {
                var _this = this;
                this.operationGrid.clearAll();
                this.operationGrid.loadData(list);
                list.forEach(( /**
                 * @param {?} item
                 * @return {?}
                 */function (item) {
                    if (_this.curSelectedKeys.indexOf(item.id) > -1) {
                        _this.operationGrid.checkRow(item.id, false);
                        if (_this.curSelected.indexOf(item) === -1) {
                            _this.curSelected.push(item);
                        }
                    }
                }));
            };
        /**
         * 构造树
         */
        /**
         * 构造树
         * @private
         * @param {?} data
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.makeTree = /**
         * 构造树
         * @private
         * @param {?} data
         * @return {?}
         */
            function (data) {
                var _this = this;
                /** @type {?} */
                var root = data.filter(( /**
                 * @param {?} t
                 * @return {?}
                 */function (t) { return t.layer === 1; })).map(( /**
                 * @param {?} t
                 * @return {?}
                 */function (t) {
                    return {
                        data: t,
                        children: [],
                        expanded: false,
                        selectable: true
                    };
                }));
                root.forEach(( /**
                 * @param {?} e
                 * @return {?}
                 */function (e) {
                    // 构造下级：children，父级的id信息
                    e.children = _this.makeChildrenByParentId(data, e.data.id);
                }));
                // 默认展开第一个节点
                root[0].expanded = true;
                return root;
            };
        /**
         * 父节点构造树
         */
        /**
         * 父节点构造树
         * @private
         * @param {?} datas
         * @param {?} parentId
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.makeChildrenByParentId = /**
         * 父节点构造树
         * @private
         * @param {?} datas
         * @param {?} parentId
         * @return {?}
         */
            function (datas, parentId) {
                var _this = this;
                /** @type {?} */
                var x = datas.filter(( /**
                 * @param {?} c
                 * @return {?}
                 */function (c) { return c.parentID === parentId; })).map(( /**
                 * @param {?} d
                 * @return {?}
                 */function (d) {
                    return {
                        data: d,
                        children: _this.makeChildrenByParentId(datas, d.id),
                        expanded: false,
                        selectable: true
                    };
                }));
                return x;
            };
        /**
         * 搜索时监测回车
         */
        /**
         * 搜索时监测回车
         * @param {?} arg
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.onKeyUp = /**
         * 搜索时监测回车
         * @param {?} arg
         * @return {?}
         */
            function (arg) {
                if (arg && arg.which === 13) {
                    this.searchData();
                }
            };
        /**
         * 搜索
         */
        /**
         * 搜索
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.searchData = /**
         * 搜索
         * @return {?}
         */
            function () {
                var _this = this;
                // 测试
                if (!this.opSearchTxt) {
                    // 全量的数据
                    this.gridData = this.gridDataSrc;
                    this.freshGridAndSelect(this.gridData);
                    return;
                }
                /** @type {?} */
                var content = this.opSearchTxt.trim();
                if (!content) {
                    // 全量的数据
                    this.gridData = this.gridDataSrc;
                    this.freshGridAndSelect(this.gridData);
                    return;
                }
                /** @type {?} */
                var filter = content.toLowerCase();
                // 如果只显示分组
                if (this.onlyShowGroup) {
                    this.gridData = this.gridDataSrc.filter(( /**
                     * @param {?} item
                     * @return {?}
                     */function (item) {
                        return item.code.toLowerCase().includes(filter) ||
                            item.name.toLowerCase().includes(filter);
                    }));
                }
                else {
                    // 需要显示操作的时候，搜索到上级时
                    this.gridData = this.gridDataSrc.filter(( /**
                     * @param {?} item
                     * @return {?}
                     */function (item) {
                        return item.code.toLowerCase().includes(filter) ||
                            item.name.toLowerCase().includes(filter) ||
                            _this.getFuncGroup(item).toLowerCase().includes(filter);
                    }));
                }
                this.freshGridAndSelect(this.gridData);
            };
        /**
         * 获取功能分组
         */
        /**
         * 获取功能分组
         * @param {?} funcOpObj
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.getFuncGroup = /**
         * 获取功能分组
         * @param {?} funcOpObj
         * @return {?}
         */
            function (funcOpObj) {
                // 获取父节点
                /** @type {?} */
                var parentId = funcOpObj.parentId;
                if (parentId) {
                    /** @type {?} */
                    var parentObj = this.gridDataSrc.filter(( /**
                     * @param {?} item
                     * @return {?}
                     */function (item) { return item.id === parentId; }));
                    if (parentObj !== null && parentObj.length > 0) {
                        return parentObj[0].name;
                    }
                }
                else {
                    return '';
                }
            };
        /**
         * 业务对象变化后
         * @param arg 业务对象参数
         */
        /**
         * 业务对象变化后
         * @param {?} arg 业务对象参数
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.bizObjectChanged = /**
         * 业务对象变化后
         * @param {?} arg 业务对象参数
         * @return {?}
         */
            function (arg) {
                if (!this.gridDataSrc || this.gridDataSrc.length === 0) {
                    return;
                }
                /** @type {?} */
                var busId = arg.node.data.id;
                // 根据业务对象过滤
                /** @type {?} */
                var lists = this.gridDataSrc.filter(( /**
                 * @param {?} item
                 * @return {?}
                 */function (item) {
                    return item.businessId === busId;
                }));
                this.freshGridAndSelect(lists);
            };
        /**
         * 获取选中数据
         */
        /**
         * 获取选中数据
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.getSelectedObject = /**
         * 获取选中数据
         * @return {?}
         */
            function () {
                /** @type {?} */
                var rows = this.operationGrid.checkeds.map(( /**
                 * @param {?} item
                 * @return {?}
                 */function (item) { return item.data; }));
                return rows;
            };
        /**
         * 选中后
         */
        /**
         * 选中后
         * @param {?} row
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.checked = /**
         * 选中后
         * @param {?} row
         * @return {?}
         */
            function (row) {
                console.log(row);
                if (row && row.id) {
                    // 是否多选
                    if (this.multiSelect) {
                        if (this.curSelectedKeys.indexOf(row.id) < 0) {
                            this.curSelected.push(row.data);
                            this.curSelectedKeys.push(row.id);
                        }
                    }
                    else {
                        this.curSelected = [row.data];
                        this.curSelectedKeys = [row.id];
                    }
                }
            };
        /**
         * 取消选中列表
         */
        /**
         * 取消选中列表
         * @param {?} row
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.unchecked = /**
         * 取消选中列表
         * @param {?} row
         * @return {?}
         */
            function (row) {
                console.log(row);
                if (row && row.id) {
                    if (this.multiSelect) {
                        /** @type {?} */
                        var x = this.curSelectedKeys.findIndex(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) { return item === row.id; }));
                        if (x >= 0) {
                            this.curSelectedKeys.splice(x, 1);
                            this.curSelected.splice(x, 1);
                        }
                    }
                    else {
                        this.curSelectedKeys = [];
                        this.curSelected = [];
                    }
                }
            };
        /**
         * 确认
         */
        /**
         * 确认
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.confirm = /**
         * 确认
         * @return {?}
         */
            function () {
                // const result = this.getSelectedObject();
                /** @type {?} */
                var result = this.curSelected;
                console.log(result);
                this.afterConfirm.next(result);
                // 取消-目的是在这里让模态框消失
                this.cancle();
            };
        /**
         * 取消
         */
        /**
         * 取消
         * @return {?}
         */
        SysFarrisFuncOperationComponent.prototype.cancle = /**
         * 取消
         * @return {?}
         */
            function () {
                this.afterCancle.next();
            };
        SysFarrisFuncOperationComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'app-sys-farris-func-operation',
                        template: "<div class=\"f-utils-absolute-all farris-split-section m-0\">\r\n    <layout [direction]=\"'h'\" [fill]=\"true\" [showBorder]=\"false\">\r\n        <layout-panel class=\"col-5 d-flex flex-column\" region=\"west\" [showBorder]=\"false\" [minWidth]=\"300\">\r\n            <div class=\"f-utils-absolute-all f-utils-flex-column farris-lsection\" style=\"background-color: white\">\r\n                <div #farrisTreeParent class=\"f-utils-fill-flex-column\" style=\"background-color: white\">   \r\n                    <farris-treetable #tt [data]=\"treeData\" [virtualized]=\"true\" [fit]=\"true\" [findField]=\"'name'\"\r\n                        [enableFindText]=\"true\" [showBorder]=\"true\" [columns]=\"treeColumns\" [idField]=\"'id'\"\r\n                        [fixedHeader]=\"true\" [showHeader]=\"true\" [remoteSort]=\"false\"\r\n                        (nodeSelected)=\"bizObjectChanged($event)\">\r\n                    </farris-treetable>\r\n                </div>\r\n            </div>\r\n        </layout-panel>\r\n        <layout-panel class=\"d-flex flex-column\" region=\"center\" [showBorder]=\"false\" [minWidth]=\"300\">\r\n            <div class=\"f-utils-absolute-all f-utils-flex-column farris-rsection\">\r\n                <form class=\"farris\u00AD-form mb-2\" style=\"width:300px;\">\r\n                    <div class=\"f-cmp-inputgroup\">\r\n                        <div class=\"input-group\"><input [(ngModel)]=\"opSearchTxt\"\r\n                                [ngModelOptions]=\"{standalone: true}\" class=\"form-control\" (keyup)=\"onKeyUp($event)\"\r\n                                placeholder=\"'\u7F16\u53F7\u6216\u8005\u540D\u79F0'\">\r\n                            <div class=\"input-group-append\"><span class=\"input-group-text\"\r\n                                    (click)=\"searchData()\">\r\n                                    <i class=\"k-icon k-i-search\"></i></span></div>\r\n                        </div>\r\n                    </div>\r\n                </form>\r\n                <!--\u5217\u8868\u4FE1\u606F-->\r\n                <div class=\"f-utils-fill-flex-column\">\r\n                    <farris-datagrid #operationGrid \r\n                    [columns]=\"gridColumns\" \r\n                    [showBorder]=\"true\"\r\n                    [virtualized]=\"true\"\r\n                    [fitColumns]=\"true\"\r\n                    [fit]=\"true\"\r\n                    [striped]=\"true\"\r\n                    [pagination]=\"false\"\r\n                    [showCheckbox]=\"true\"\r\n                    [selectOnCheck]=\"true\"\r\n                    [checkOnSelect]=\"true\"\r\n                    [multiSelect]=\"multiSelect\"\r\n                    (checked)=\"checked($event)\"\r\n                    (unChecked)=\"unchecked($event)\">\r\n                </farris-datagrid>\r\n                </div>\r\n            </div>\r\n        </layout-panel>\r\n    </layout>\r\n</div>\r\n<ng-template #btnOkAndCancle>\r\n    <button class=\"btn btn-sm btn-primary\" (click)=\"confirm()\">\u786E\u5B9A</button>\r\n    <button class=\"btn btn-sm btn-secondary\" (click)=\"cancle()\">\u53D6\u6D88</button>\r\n</ng-template>",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        SysFarrisFuncOperationComponent.ctorParameters = function () {
            return [
                { type: SysFuncOperationService },
                { type: BizObjectService },
                { type: SysLocaleService }
            ];
        };
        SysFarrisFuncOperationComponent.propDecorators = {
            bizObjectId: [{ type: i0.Input }],
            gridSelectMode: [{ type: i0.Input }],
            sessionId: [{ type: i0.Input }],
            onlyShowGroup: [{ type: i0.Input }],
            isGroupCanSelected: [{ type: i0.Input }],
            afterConfirm: [{ type: i0.Output }],
            afterCancle: [{ type: i0.Output }],
            btnOkAndCancleRef: [{ type: i0.ViewChild, args: ['btnOkAndCancle',] }],
            operationGrid: [{ type: i0.ViewChild, args: ['operationGrid',] }],
            farrisTreeParent: [{ type: i0.ViewChild, args: ['farrisTreeParent',] }],
            farrisTt: [{ type: i0.ViewChild, args: ['tt',] }],
            winResize: [{ type: i0.HostListener, args: ['window:resize', [],] }]
        };
        return SysFarrisFuncOperationComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SysFuncOpHelpComponent = /** @class */ (function () {
        function SysFuncOpHelpComponent(modalService, injector, cfr, sysUtil, localService) {
            this.modalService = modalService;
            this.injector = injector;
            this.cfr = cfr;
            this.sysUtil = sysUtil;
            this.localService = localService;
            /**
             * 是否启用新UI,启用后性能提升
             */
            this.useNewUI = false;
            /**
             * 初始选中的权限对象内码
             */
            this.initSelectKeys = [];
            /**
             * 是否只显示分组
             */
            this.onlyShowGroup = false;
            /**
             * 输出事件
             */
            this.afterHelpConfirm = new i0.EventEmitter();
            if (this.localService) {
                this.i18nRes = this.localService.getComponentOpt('help');
            }
            else {
                this.i18nRes = HELP_LOCALE_ZHCHS;
            }
        }
        /**
         * @return {?}
         */
        SysFuncOpHelpComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
            };
        /**
         * 新UI
         */
        /**
         * 新UI
         * @param {?} sessionId
         * @return {?}
         */
        SysFuncOpHelpComponent.prototype.showNewHelp = /**
         * 新UI
         * @param {?} sessionId
         * @return {?}
         */
            function (sessionId) {
                var _this = this;
                // 调用帮助时必须传递sessionId
                this.sysUtil.sessionId = sessionId;
                /** @type {?} */
                var compFactory = this.cfr.resolveComponentFactory(SysFarrisFuncOperationComponent);
                /** @type {?} */
                var funcOpCompNew = compFactory.create(this.injector);
                // 获取功能操作的实例组件
                if (this.initSelectKeys != null && this.initSelectKeys.length > 0) {
                    this.initSelectKeys.forEach(( /**
                     * @param {?} key
                     * @return {?}
                     */function (key) { return funcOpCompNew.instance.curSelectedKeys.push(key); }));
                }
                // 初始化选择的业务操作
                if (this.bizObjectId && this.bizObjectId !== '') {
                    funcOpCompNew.instance.bizObjectId = this.bizObjectId;
                }
                if (this.gridSelectMode) {
                    funcOpCompNew.instance.gridSelectMode = this.gridSelectMode;
                }
                funcOpCompNew.instance.onlyShowGroup = this.onlyShowGroup;
                funcOpCompNew.instance.isGroupCanSelected = this.isGroupCanSelected;
                /** @type {?} */
                var modalConfig = {
                    title: this.i18nRes.operationTitle,
                    width: 850,
                    height: 550,
                    buttons: funcOpCompNew.instance.btnOkAndCancleRef,
                    showButtons: true,
                    showMaxButton: true,
                    resizable: true
                };
                // 弹出帮助框
                this.funcOpDialog = this.modalService.show(funcOpCompNew, modalConfig);
                // 取消后
                funcOpCompNew.instance.afterCancle.subscribe(( /**
                 * @param {?} res
                 * @return {?}
                 */function (res) {
                    // 消失模态框
                    _this.funcOpDialog.close();
                }));
                // 确认后
                funcOpCompNew.instance.afterConfirm.subscribe(( /**
                 * @param {?} res
                 * @return {?}
                 */function (res) {
                    _this.afterHelpConfirm.next(res);
                }));
            };
        /**
         * 弹出Farris模态框帮助
         */
        /**
         * 弹出Farris模态框帮助
         * @param {?} sessionId
         * @return {?}
         */
        SysFuncOpHelpComponent.prototype.showHelp = /**
         * 弹出Farris模态框帮助
         * @param {?} sessionId
         * @return {?}
         */
            function (sessionId) {
                var _this = this;
                if (this.useNewUI) {
                    this.showNewHelp(sessionId);
                    return;
                }
                // 调用帮助时必须传递sessionId
                this.sysUtil.sessionId = sessionId;
                /** @type {?} */
                var compFactory = this.cfr.resolveComponentFactory(SysFuncOperationComponent);
                this.funcOpComp = compFactory.create(this.injector);
                // 获取功能操作的实例组件
                if (this.initSelectKeys != null && this.initSelectKeys.length > 0) {
                    this.initSelectKeys
                        .forEach(( /**
                 * @param {?} key
                 * @return {?}
                 */function (key) { return _this.funcOpComp.instance.curSelectedKeys.push(key); }));
                }
                // 初始化选择的业务操作
                if (this.bizObjectId && this.bizObjectId !== '') {
                    this.funcOpComp.instance.bizObjectId = this.bizObjectId;
                }
                if (this.gridSelectMode) {
                    this.funcOpComp.instance.gridSelectMode = this.gridSelectMode;
                }
                this.funcOpComp.instance.onlyShowGroup = this.onlyShowGroup;
                this.funcOpComp.instance.isGroupCanSelected = this.isGroupCanSelected;
                // 旧版本的需要在这里loaddata
                this.funcOpComp.instance.loadData();
                /** @type {?} */
                var modalConfig = {
                    title: '功能操作帮助',
                    width: 800,
                    height: 550,
                    buttons: this.funcOpComp.instance.btnOkAndCancleRef,
                    showButtons: true,
                    showMaxButton: false,
                    resizable: false
                };
                // 弹出帮助框
                this.funcOpDialog = this.modalService.show(this.funcOpComp, modalConfig);
                // 取消后
                this.funcOpComp.instance.afterCancle.subscribe(( /**
                 * @param {?} res
                 * @return {?}
                 */function (res) {
                    // 消失模态框
                    _this.funcOpDialog.close();
                }));
                // 确认后
                this.funcOpComp.instance.afterConfirm.subscribe(( /**
                 * @param {?} res
                 * @return {?}
                 */function (res) {
                    _this.afterHelpConfirm.next(res);
                }));
            };
        SysFuncOpHelpComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'app-sys-func-op-help',
                        template: "",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        SysFuncOpHelpComponent.ctorParameters = function () {
            return [
                { type: uiModal.BsModalService },
                { type: i0.Injector },
                { type: i0.ComponentFactoryResolver },
                { type: SysFrontUtil },
                { type: SysLocaleService }
            ];
        };
        SysFuncOpHelpComponent.propDecorators = {
            bizObjectId: [{ type: i0.Input }],
            useNewUI: [{ type: i0.Input }],
            initSelectKeys: [{ type: i0.Input }],
            onlyShowGroup: [{ type: i0.Input }],
            gridSelectMode: [{ type: i0.Input }],
            isGroupCanSelected: [{ type: i0.Input }],
            afterHelpConfirm: [{ type: i0.Output }]
        };
        return SysFuncOpHelpComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BizObjectComponent = /** @class */ (function () {
        function BizObjectComponent(bizObjectService) {
            this.bizObjectService = bizObjectService;
            this.selectKeys = [];
            /**
             * 业务对象变化后
             */
            this.bizObjectChanged = new i0.EventEmitter();
            /**
             * 树控件展开的节点值
             */
            this.expandedKeys = [];
        }
        /**
         * @return {?}
         */
        BizObjectComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.bizObjectService.getBizObjects().subscribe(( /**
                 * @param {?} result
                 * @return {?}
                 */function (result) {
                    result.forEach(( /**
                     * @param {?} element
                     * @return {?}
                     */function (element) {
                        if (!element.parentID) {
                            element.parentID = undefined;
                            // 展开第一级
                            if (element.layer === 1) {
                                _this.expandedKeys.push(element.id);
                            }
                        }
                    }));
                    _this.data = result;
                    // 默认业务对象的处理
                    if (_this.bizObjectId && _this.bizObjectId !== '') {
                        // 树焦点行
                        /** @type {?} */
                        var curBizObjectId_1 = _this.bizObjectId;
                        while (curBizObjectId_1) {
                            _this.expandedKeys.push(curBizObjectId_1);
                            // 找到当前节点的父节点
                            curBizObjectId_1 = _this.data.find(( /**
                             * @param {?} item
                             * @return {?}
                             */function (item) { return item.id === curBizObjectId_1; })).parentID;
                        }
                        // 右侧操作变化
                        _this.selectKeys = [_this.bizObjectId];
                        setTimeout(( /**
                         * @return {?}
                         */function () {
                            _this.bizObjectChanged.emit(_this.bizObjectId);
                        }), 80);
                    }
                }));
            };
        /** 选中行改变后触发事件 */
        /**
         * 选中行改变后触发事件
         * @param {?} args
         * @return {?}
         */
        BizObjectComponent.prototype.selectChange = /**
         * 选中行改变后触发事件
         * @param {?} args
         * @return {?}
         */
            function (args) {
                /** @type {?} */
                var curBusId = args.dataItem.id;
                // 首次触发
                if (this.selectKeys == null || this.selectKeys.length === 0) {
                    this.bizObjectChanged.emit(curBusId);
                }
                else {
                    // 这个事件里还不会对selectedkeys数组赋值，是上一个状态的
                    /** @type {?} */
                    var lastBusId = this.selectKeys[0];
                    // 变化的情况下触发
                    if (lastBusId !== curBusId) {
                        this.bizObjectChanged.emit(curBusId);
                    }
                }
            };
        BizObjectComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'app-biz-object',
                        template: "<div class=\"d-flex flex-column flex-fill pt-2 pb-2\" style=\"height: 100%\">\r\n    <kendo-treeview [nodes]=\"data\" textField=\"name\" kendoTreeViewExpandable kendoTreeViewFlatDataBinding idField=\"id\"\r\n      parentIdField=\"parentID\" kendoTreeViewSelectable [selectBy]=\"'id'\" [(selectedKeys)]=\"this.selectKeys\"\r\n      [(expandedKeys)]=\"expandedKeys\"\r\n      [expandBy]=\"'id'\"\r\n      (selectionChange)=\"selectChange($event)\">\r\n    </kendo-treeview>\r\n</div>",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        BizObjectComponent.ctorParameters = function () {
            return [
                { type: BizObjectService }
            ];
        };
        BizObjectComponent.propDecorators = {
            bizObjectId: [{ type: i0.Input }],
            bizObjectChanged: [{ type: i0.Output }]
        };
        return BizObjectComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SysFuncOpHelpModule = /** @class */ (function () {
        function SysFuncOpHelpModule() {
        }
        SysFuncOpHelpModule.decorators = [
            { type: i0.NgModule, args: [{
                        imports: [
                            common.CommonModule,
                            i1.HttpClientModule,
                            forms.FormsModule,
                            forms.ReactiveFormsModule,
                            uiDatagrid.DatagridModule,
                            uiTreetable.TreeTableModule,
                            uiLayout.LayoutModule,
                            kendoAngularTreeview.TreeViewModule,
                            kendoAngularLayout.LayoutModule,
                            uiLayout.LayoutModule,
                            kendoAngularGrid.GridModule,
                            uiModal.ModalModule.forRoot(),
                            uiMessager.MessagerModule.forRoot()
                        ],
                        declarations: [
                            SysFuncOperationComponent,
                            SysFuncOpHelpComponent,
                            BizObjectComponent,
                            SysFarrisFuncOperationComponent
                        ],
                        providers: [
                            SysFrontUtil,
                            SysLocaleService,
                            SysFuncOperationService,
                            BizObjectService
                        ],
                        entryComponents: [
                            SysFuncOperationComponent,
                            SysFarrisFuncOperationComponent
                        ],
                        exports: [
                            SysFuncOperationComponent,
                            SysFuncOpHelpComponent,
                            BizObjectComponent,
                            SysFarrisFuncOperationComponent
                        ]
                    },] }
        ];
        return SysFuncOpHelpModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SysHelpService = /** @class */ (function () {
        function SysHelpService(http, sysUtil) {
            this.http = http;
            this.sysUtil = sysUtil;
        }
        /**
         * @private
         * @param {?} filter
         * @return {?}
         */
        SysHelpService.prototype.getHttpOption = /**
         * @private
         * @param {?} filter
         * @return {?}
         */
            function (filter) {
                i2.SysDataUtil.deleteEmpty(filter);
                /** @type {?} */
                var paramStr = JSON.stringify(filter);
                /** @type {?} */
                var opt = {
                    headers: this.sysUtil.getHeader()['headers'],
                    params: { param: paramStr }
                };
                return opt;
            };
        /**
         * @param {?} filter
         * @return {?}
         */
        SysHelpService.prototype.getUsers = /**
         * @param {?} filter
         * @return {?}
         */
            function (filter) {
                /** @type {?} */
                var opt = this.getHttpOption(filter);
                // 用户目前的有分页信息
                return this.http.get(this.sysUtil.userWebApi, opt).pipe(operators.map(( /**
                 * @param {?} val
                 * @return {?}
                 */function (val) {
                    return val.data;
                })));
            };
        /**
         * @param {?} filter
         * @return {?}
         */
        SysHelpService.prototype.getPositions = /**
         * @param {?} filter
         * @return {?}
         */
            function (filter) {
                /** @type {?} */
                var opt = this.getHttpOption(filter);
                return this.http.get(this.sysUtil.positionWebApi, opt).pipe(operators.map(( /**
                 * @param {?} val
                 * @return {?}
                 */function (val) {
                    return val.data;
                })));
            };
        /**
         * @param {?} filter
         * @return {?}
         */
        SysHelpService.prototype.getRoles = /**
         * @param {?} filter
         * @return {?}
         */
            function (filter) {
                /** @type {?} */
                var opt = this.getHttpOption(filter);
                return this.http.get(this.sysUtil.roleWebApi, opt).pipe(operators.map(( /**
                 * @param {?} val
                 * @return {?}
                 */function (val) {
                    return val.data;
                })));
            };
        /**
         * @param {?} filter
         * @return {?}
         */
        SysHelpService.prototype.getOrgs = /**
         * @param {?} filter
         * @return {?}
         */
            function (filter) {
                /** @type {?} */
                var opt = this.getHttpOption(filter);
                // 视图
                opt.params.useView = true;
                return this.handerTreeData(this.http.get(this.sysUtil.sysOrgWebApi + '/all', opt));
            };
        /**
         * @param {?} filter
         * @return {?}
         */
        SysHelpService.prototype.getUserGroups = /**
         * @param {?} filter
         * @return {?}
         */
            function (filter) {
                /** @type {?} */
                var opt = this.getHttpOption(filter);
                return this.handerTreeData(this.http.get(this.sysUtil.userGroupWebApi + '/all', opt));
            };
        /**
         * @param {?} filter
         * @return {?}
         */
        SysHelpService.prototype.getPositionGroups = /**
         * @param {?} filter
         * @return {?}
         */
            function (filter) {
                /** @type {?} */
                var opt = this.getHttpOption(filter);
                return this.handerTreeData(this.http.get(this.sysUtil.posGroupWebApi + '/all', opt));
            };
        /**
         * @param {?} filter
         * @return {?}
         */
        SysHelpService.prototype.getRoleGroups = /**
         * @param {?} filter
         * @return {?}
         */
            function (filter) {
                /** @type {?} */
                var opt = this.getHttpOption(filter);
                return this.handerTreeData(this.http.get(this.sysUtil.roleGroupWebApi + '/all', opt));
            };
        /**
         * @param {?} item
         * @return {?}
         */
        SysHelpService.prototype.convert2Json = /**
         * @param {?} item
         * @return {?}
         */
            function (item) {
                /** @type {?} */
                var result = {
                    id: item.id,
                    code: item.code,
                    name: item.name,
                    parentId: item.parentId || '',
                    path: item.path,
                    layer: item.layer,
                    isDetail: item.isDetail,
                };
                if (item.nodeType != null && item.nodeType !== undefined) {
                    result.nodeType = item.nodeType;
                }
                return result;
            };
        /**
         * @param {?} org
         * @return {?}
         */
        SysHelpService.prototype.getParentOrg = /**
         * @param {?} org
         * @return {?}
         */
            function (org) {
                /** @type {?} */
                var paramStr = JSON.stringify(this.convert2Json(org));
                /** @type {?} */
                var opt = {
                    headers: this.sysUtil.getHeader()['headers'],
                    params: { param: paramStr }
                };
                return this.http.get(this.sysUtil.sysOrgWebApi + '/getparentorg/0002', opt).pipe(operators.map(( /**
                 * @param {?} val
                 * @return {?}
                 */function (val) {
                    return val;
                })));
            };
        /** 适配Kendo-treeView结构 */
        /**
         * 适配Kendo-treeView结构
         * @private
         * @param {?} src
         * @return {?}
         */
        SysHelpService.prototype.handerTreeData = /**
         * 适配Kendo-treeView结构
         * @private
         * @param {?} src
         * @return {?}
         */
            function (src) {
                return src.pipe(operators.map(( /**
                 * @param {?} val
                 * @return {?}
                 */function (val) {
                    rxjs.from(val).subscribe(( /**
                     * @param {?} item
                     * @return {?}
                     */function (item) {
                        if (item.layer == 1 || item.parentId == null || item.parentId === '') {
                            item.parentId = undefined;
                        }
                    }));
                    return val;
                })));
            };
        SysHelpService.decorators = [
            { type: i0.Injectable }
        ];
        /** @nocollapse */
        SysHelpService.ctorParameters = function () {
            return [
                { type: i1.HttpClient },
                { type: i2.SysCommonUtilService }
            ];
        };
        return SysHelpService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SysExceptionService = /** @class */ (function () {
        function SysExceptionService() {
        }
        /**
         * @param {?} error
         * @return {?}
         */
        SysExceptionService.prototype.getErrorInfo = /**
         * @param {?} error
         * @return {?}
         */
            function (error) {
                return error.Message;
            };
        SysExceptionService.decorators = [
            { type: i0.Injectable }
        ];
        return SysExceptionService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var GroupService = /** @class */ (function () {
        function GroupService(http, sysUtil) {
            this.http = http;
            this.sysUtil = sysUtil;
        }
        /**
         * @private
         * @param {?} type
         * @return {?}
         */
        GroupService.prototype.getUrlByGroupType = /**
         * @private
         * @param {?} type
         * @return {?}
         */
            function (type) {
                /** @type {?} */
                var url = '';
                switch (type) {
                    case 'user':
                        url = this.sysUtil.userGroupWebApi;
                        break;
                    case 'position':
                        url = this.sysUtil.posGroupWebApi;
                        break;
                    case 'role':
                        url = this.sysUtil.roleGroupWebApi;
                        break;
                    case 'org':
                        url = this.sysUtil.sysOrgWebApi;
                        break;
                    // 默认用户分组
                    default:
                        url = this.sysUtil.userGroupWebApi;
                        break;
                }
                return url;
            };
        /**
         * 根据级数获取组织
         * @param layer 级数
         */
        /**
         * 根据级数获取组织
         * @param {?} layer 级数
         * @param {?} type
         * @param {?} filter
         * @return {?}
         */
        GroupService.prototype.getByLayer = /**
         * 根据级数获取组织
         * @param {?} layer 级数
         * @param {?} type
         * @param {?} filter
         * @return {?}
         */
            function (layer, type, filter) {
                /** @type {?} */
                var url = this.getUrlByGroupType(type);
                if (!filter) {
                    filter = {};
                }
                filter.layer = layer;
                filter.parentId = null;
                i2.SysDataUtil.deleteEmpty(filter);
                /** @type {?} */
                var opt = {
                    headers: this.sysUtil.getHeader().headers,
                    params: { param: JSON.stringify(filter) }
                };
                return this.http.get(url, opt);
            };
        /**
         * 根据级数获取组织
         * @param parentId 父节点
         */
        /**
         * 根据级数获取组织
         * @param {?} parentId 父节点
         * @param {?} type
         * @param {?} filter
         * @return {?}
         */
        GroupService.prototype.getByParentId = /**
         * 根据级数获取组织
         * @param {?} parentId 父节点
         * @param {?} type
         * @param {?} filter
         * @return {?}
         */
            function (parentId, type, filter) {
                /** @type {?} */
                var url = this.getUrlByGroupType(type);
                if (!filter) {
                    filter = {};
                }
                filter.layer = null;
                filter.parentId = parentId;
                i2.SysDataUtil.deleteEmpty(filter);
                /** @type {?} */
                var opt = {
                    headers: this.sysUtil.getHeader().headers,
                    params: { param: JSON.stringify(filter) }
                };
                return this.http.get(url, opt);
            };
        /**
         * 根据级数获取组织
         * @param parentId 父节点
         */
        /**
         * 根据级数获取组织
         * @param {?} type
         * @param {?} filter
         * @return {?}
         */
        GroupService.prototype.searchData = /**
         * 根据级数获取组织
         * @param {?} type
         * @param {?} filter
         * @return {?}
         */
            function (type, filter) {
                /** @type {?} */
                var url = this.getUrlByGroupType(type);
                if (!filter) {
                    filter = {};
                }
                filter.layer = null;
                filter.parentId = null;
                i2.SysDataUtil.deleteEmpty(filter);
                /** @type {?} */
                var opt = {
                    headers: this.sysUtil.getHeader().headers,
                    params: { param: JSON.stringify(filter) }
                };
                return this.http.get(url, opt);
            };
        GroupService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */
        GroupService.ctorParameters = function () {
            return [
                { type: i1.HttpClient },
                { type: i2.SysCommonUtilService }
            ];
        };
        /** @nocollapse */ GroupService.ngInjectableDef = i0.defineInjectable({ factory: function GroupService_Factory() { return new GroupService(i0.inject(i1.HttpClient), i0.inject(i2.SysCommonUtilService)); }, token: GroupService, providedIn: "root" });
        return GroupService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var GroupNavComponent = /** @class */ (function () {
        function GroupNavComponent(groupService, localService, notifyService, msg) {
            this.groupService = groupService;
            this.localService = localService;
            this.notifyService = notifyService;
            this.msg = msg;
            this.showHeader = true;
            this.fixedHeader = true;
            this.withCodeColumn = true;
            // 是否添加所有根节点
            this.allRoot = false;
            this.selectKeys = [];
            this.needRemote = true;
            /**
             * 显示组织类型列
             */
            this.showNodeTypeColumn = false;
            this.showFilterBar = false;
            this.searchFields = [{ label: "编号", value: "code" }, { label: "名称", value: "name" }];
            this.allColumnsTitle = '所有列';
            this.search = new i0.EventEmitter();
            this.selectRowChanged = new i0.EventEmitter();
            /**
             * 查询内容
             */
            this.searchInfo = '';
            /**
             * 上次查询内容
             */
            this.oldSearchInfo = '';
            this.field = '';
            /**
             * 上次查询内容
             */
            this.oldField = '';
            //上次搜索结果
            this.searchResult = [];
            //上次搜索的位置
            this.searchIndex = 0;
            /**
             * 监控拖拽事件，重绘组织TreeTable
             * @param event 事件参数
             */
            this.temp = true;
        }
        /**
         * @return {?}
         */
        GroupNavComponent.prototype.winResize = /**
         * @return {?}
         */
            function () {
                this.reSizeTreeTable();
            };
        /** 重绘Farris-TreeTable的大小 */
        /**
         * 重绘Farris-TreeTable的大小
         * @param {?=} height
         * @param {?=} width
         * @return {?}
         */
        GroupNavComponent.prototype.reSizeTreeTable = /**
         * 重绘Farris-TreeTable的大小
         * @param {?=} height
         * @param {?=} width
         * @return {?}
         */
            function (height, width) {
                if (height === void 0) {
                    height = 0;
                }
                if (width === void 0) {
                    width = 0;
                }
                if (height === 0) {
                    height = this.farrisTreeParent.nativeElement.clientHeight;
                }
                if (width === 0) {
                    width = this.farrisTreeParent.nativeElement.clientWidth;
                }
                this.farrisTt.resize({ height: height, width: width });
            };
        /**
         * @return {?}
         */
        GroupNavComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                if (this.localService) {
                    this.i18nRes = this.localService.getComponentOpt('navi');
                }
                else {
                    this.i18nRes = NAVI_LOCALE_ZHCHS;
                }
                this.searchFields = [{ label: this.i18nRes.codeField, value: "code" }, { label: this.i18nRes.nameField, value: "name" }];
                this.allColumnsTitle = this.i18nRes.allColumnsTitle;
                // 启用编号列
                if (this.withCodeColumn) {
                    this.fixedHeader = true;
                    this.showHeader = true;
                    this.cols = [
                        { field: 'code', title: this.i18nRes.codeField, width: 100 },
                        { field: 'name', title: this.i18nRes.nameField, width: 100 }
                    ];
                }
                else {
                    this.fixedHeader = false;
                    this.showHeader = false;
                    this.cols = [
                        { field: 'name', title: this.i18nRes.nameField, width: 100 }
                    ];
                }
                // 需要获取数据，否则直接根据treeData初始数据
                if (this.needRemote) {
                    // 所有标识
                    if (this.allRoot) {
                        this.treeData = [];
                        /** @type {?} */
                        var rootNode = {};
                        rootNode.children = [];
                        rootNode.data = {
                            id: 'all',
                            code: this.i18nRes.rootNodeCode,
                            name: this.i18nRes.rootNodeTitle
                        };
                        rootNode.expanded = true;
                        rootNode.leaf = false;
                        rootNode.selectable = true;
                        this.treeData.push(rootNode);
                    }
                    else {
                        this.treeData = [];
                    }
                    this.groupService.getByLayer(1, this.groupType, this.filter).subscribe(( /**
                     * @param {?} val
                     * @return {?}
                     */function (val) {
                        _this.initTreeNodeList(val);
                    }));
                }
                else { // 不需要从服务器端获取
                    this.farrisTt.updateSerializedValue();
                }
                // 异步加载，通过children为0来判断从未展开过节点
                this.farrisTt.expand.pipe(operators.filter(( /**
                 * @param {?} node
                 * @return {?}
                 */function (node) { return node.children.length === 0; }))).subscribe(( /**
                 * @param {?} node
                 * @return {?}
                 */function (node) {
                    node['showLoading'] = true;
                    /** @type {?} */
                    var id = node.data.id;
                    // 获取系统组织
                    _this.groupService.getByParentId(id, _this.groupType, _this.filter).subscribe(( /**
                     * @param {?} list
                     * @return {?}
                     */function (list) {
                        node['showLoading'] = false;
                        /** @type {?} */
                        var childNode = [];
                        // 转换下级组织
                        rxjs.from(list).subscribe(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) {
                            childNode.push(_this.convertToTreeNode(item));
                        }));
                        // 防止出现下级节点没有数据
                        if (childNode.length > 0) {
                            _this.farrisTt.appendChildren(childNode, node);
                        }
                        else { // 应该是无下级权限才走这里
                            node.leaf = true;
                            _this.farrisTt.updateSerializedValue();
                        }
                    }));
                }));
            };
        /**
         * 刷新当前页面信息
         */
        /**
         * 刷新当前页面信息
         * @return {?}
         */
        GroupNavComponent.prototype.refreshThisNav = /**
         * 刷新当前页面信息
         * @return {?}
         */
            function () {
                var _this = this;
                this.treeData = [];
                /** @type {?} */
                var rootNode = {};
                rootNode.children = [];
                rootNode.data = {
                    id: 'all',
                    code: this.i18nRes.rootNodeCode,
                    name: this.i18nRes.rootNodeTitle
                };
                rootNode.expanded = true;
                rootNode.leaf = false;
                rootNode.selectable = true;
                this.treeData.push(rootNode);
                this.groupService.getByLayer(1, this.groupType, this.filter).subscribe(( /**
                 * @param {?} val
                 * @return {?}
                 */function (val) {
                    _this.initTreeNodeList(val);
                }));
            };
        /** 页面渲染完成后 */
        /**
         * 页面渲染完成后
         * @return {?}
         */
        GroupNavComponent.prototype.ngAfterContentInit = /**
         * 页面渲染完成后
         * @return {?}
         */
            function () {
                this.reSizeTreeTable();
            };
        /** 搜索， 此处是分层加载，搜索时需要根据原数据进行搜索，搜索后，需要将对应父节点展示
         *   @param args  格式 searchData = { field: '*', value: '' };
         */
        /**
         * 搜索， 此处是分层加载，搜索时需要根据原数据进行搜索，搜索后，需要将对应父节点展示
         * @param {?} args  格式 searchData = { field: '*', value: '' };
         * @return {?}
         */
        GroupNavComponent.prototype.searchData = /**
         * 搜索， 此处是分层加载，搜索时需要根据原数据进行搜索，搜索后，需要将对应父节点展示
         * @param {?} args  格式 searchData = { field: '*', value: '' };
         * @return {?}
         */
            function (args) {
                var _this = this;
                this.search.next(args);
                if (!this.filter) {
                    this.filter = {};
                }
                if (!args) {
                    return;
                }
                this.searchInfo = args.value;
                this.field = args.field;
                // 搜索内容为空
                if (!this.searchInfo || this.searchInfo.length <= 0) {
                    this.clearSearchValue();
                    return;
                }
                // 搜索内容与上次不同，重新搜索
                if (this.oldSearchInfo !== this.searchInfo || this.oldField !== this.field) {
                    this.filter.isSearchTree = true;
                    if (this.field === '*') {
                        this.filter.codeOrName = this.searchInfo;
                        this.filter.code = '';
                        this.filter.name = '';
                    }
                    else if (this.field === 'code') {
                        this.filter.code = this.searchInfo;
                        this.filter.name = '';
                        this.filter.codeOrName = '';
                    }
                    else if (this.field === 'name') {
                        this.filter.name = this.searchInfo;
                        this.filter.code = '';
                        this.filter.codeOrName = '';
                    }
                    this.groupService.searchData(this.groupType, this.filter).subscribe(( /**
                     * @param {?} list
                     * @return {?}
                     */function (list) {
                        _this.treeData = [];
                        _this.searchResult = [];
                        _this.searchIndex = 0;
                        if (list && list.length > 0) {
                            _this.treeData = _this.makeTree(list);
                            list.forEach(( /**
                             * @param {?} item
                             * @return {?}
                             */function (item) {
                                if (item.search_data) {
                                    _this.searchResult.push(item.id);
                                }
                            }));
                            // 手动更新TreeTable的值
                            _this.farrisTt.updateSerializedValue();
                            setTimeout(( /**
                             * @return {?}
                             */function () {
                                _this.farrisTt.selectNode(_this.searchResult[0]);
                                _this.farrisTt.collapseAll();
                                _this.expandParent(_this.searchResult[0]);
                                _this.farrisTt.scrollToCurrentNode();
                            }));
                        }
                        else {
                            // 手动更新TreeTable的值
                            _this.farrisTt.updateSerializedValue();
                            _this.notifyService.info(_this.i18nRes.searchNullData);
                        }
                    }));
                }
                else {
                    if (this.searchResult && this.searchResult.length > 0) {
                        this.searchIndex = this.searchIndex + 1;
                        // 已经搜索到最后，需要重头搜索
                        if (this.searchIndex === this.searchResult.length) {
                            this.searchIndex = 0;
                        }
                        setTimeout(( /**
                         * @return {?}
                         */function () {
                            _this.farrisTt.selectNode(_this.searchResult[_this.searchIndex]);
                            _this.farrisTt.collapseAll();
                            _this.expandParent(_this.searchResult[_this.searchIndex]);
                            _this.farrisTt.scrollToCurrentNode();
                        }));
                    }
                    else {
                        this.notifyService.info(this.i18nRes.searchNullData);
                    }
                }
                this.oldSearchInfo = this.searchInfo;
                this.oldField = this.field;
            };
        /**
         * @return {?}
         */
        GroupNavComponent.prototype.clearSearchValue = /**
         * @return {?}
         */
            function () {
                var _this = this;
                //点击清除时，防止事件执行两次
                if (!this.temp) {
                    return;
                }
                this.temp = false;
                // 所有标识
                if (this.allRoot) {
                    this.treeData = [];
                    /** @type {?} */
                    var rootNode = {};
                    rootNode.children = [];
                    rootNode.data = {
                        id: 'all',
                        code: this.i18nRes.rootNodeCode,
                        name: this.i18nRes.rootNodeTitle
                    };
                    rootNode.expanded = true;
                    rootNode.leaf = false;
                    rootNode.selectable = true;
                    this.treeData.push(rootNode);
                }
                else {
                    this.treeData = [];
                }
                this.filter.isSearchTree = false;
                this.filter.codeOrName = null;
                this.filter.code = null;
                this.filter.name = null;
                this.groupService.getByLayer(1, this.groupType, this.filter).subscribe(( /**
                 * @param {?} val
                 * @return {?}
                 */function (val) {
                    _this.initTreeNodeList(val);
                    if (_this.treeData.length > 0) {
                        _this.farrisTt.selectNode(_this.treeData[0].data.id);
                    }
                    _this.temp = true;
                }));
                this.oldSearchInfo = '';
                this.oldField = '';
            };
        /**
         * 构造树
        */
        /**
         * 构造树
         * @private
         * @param {?} data
         * @return {?}
         */
        GroupNavComponent.prototype.makeTree = /**
         * 构造树
         * @private
         * @param {?} data
         * @return {?}
         */
            function (data) {
                var _this = this;
                /** @type {?} */
                var root = data.filter(( /**
                 * @param {?} t
                 * @return {?}
                 */function (t) { return t.layer === 1; })).map(( /**
                 * @param {?} t
                 * @return {?}
                 */function (t) {
                    /** @type {?} */
                    var node = {
                        data: t,
                        children: [],
                        expanded: false,
                        selectable: t.farris_selectable
                    };
                    if (!t.isDetail) {
                        node.leaf = false;
                        node.children = [];
                    }
                    return node;
                }));
                root.forEach(( /**
                 * @param {?} e
                 * @return {?}
                 */function (e) {
                    // 构造下级：children，父级的id信息
                    e.children = _this.makeChildrenByParentId(data, e.data.id);
                }));
                return root;
            };
        /**
         * 父节点构造树
         */
        /**
         * 父节点构造树
         * @private
         * @param {?} datas
         * @param {?} parentId
         * @return {?}
         */
        GroupNavComponent.prototype.makeChildrenByParentId = /**
         * 父节点构造树
         * @private
         * @param {?} datas
         * @param {?} parentId
         * @return {?}
         */
            function (datas, parentId) {
                var _this = this;
                /** @type {?} */
                var x = datas.filter(( /**
                 * @param {?} c
                 * @return {?}
                 */function (c) { return c.parentId === parentId; })).map(( /**
                 * @param {?} d
                 * @return {?}
                 */function (d) {
                    return {
                        data: d,
                        children: _this.makeChildrenByParentId(datas, d.id),
                        expanded: false,
                        selectable: d.farris_selectable
                    };
                }));
                return x;
            };
        /**
         * 将组织实体转换为Farris-TreeNode
         */
        /**
         * 将组织实体转换为Farris-TreeNode
         * @private
         * @param {?} src
         * @return {?}
         */
        GroupNavComponent.prototype.convertToTreeNode = /**
         * 将组织实体转换为Farris-TreeNode
         * @private
         * @param {?} src
         * @return {?}
         */
            function (src) {
                /** @type {?} */
                var node = { data: src };
                // 非明细节点
                if (!src.isDetail) {
                    node.leaf = false;
                    node.children = [];
                }
                // TODO: 只有有权限的组织才能被选中,还是这里不做处理，发送事件时去处理
                node.selectable = src.farris_selectable;
                return node;
            };
        /**
         * 初始顶级分组，默认layer=1的情况
         * @param sysOrgList 系统组织列表
         */
        /**
         * 初始顶级分组，默认layer=1的情况
         * @private
         * @param {?} list
         * @return {?}
         */
        GroupNavComponent.prototype.initTreeNodeList = /**
         * 初始顶级分组，默认layer=1的情况
         * @private
         * @param {?} list
         * @return {?}
         */
            function (list) {
                var _this = this;
                if (!list || list.length === 0) {
                    return;
                }
                rxjs.from(list).subscribe(( /**
                 * @param {?} firstLayer
                 * @return {?}
                 */function (firstLayer) {
                    /** @type {?} */
                    var node = _this.convertToTreeNode(firstLayer);
                    // 非明细，带往下多构造一级，假设有数据
                    if (node.leaf) {
                        /** @type {?} */
                        var curId_1 = firstLayer.id;
                        /** @type {?} */
                        var childern = list.filter(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) { return item.parentId === curId_1; }));
                        rxjs.from(childern).subscribe(( /**
                         * @param {?} secondLayer
                         * @return {?}
                         */function (secondLayer) {
                            node.children.push(_this.convertToTreeNode(secondLayer));
                        }));
                    }
                    // 如果添加所有根节点
                    if (_this.allRoot) {
                        _this.treeData[0].children.push(node);
                    }
                    else {
                        _this.treeData.push(node);
                    }
                }));
                // 手动更新TreeTable的值
                this.farrisTt.updateSerializedValue();
            };
        /**
         * 选择的组织行
         * @param arg 参数
         */
        /**
         * 选择的组织行
         * @param {?} event
         * @return {?}
         */
        GroupNavComponent.prototype.onNodeSelect = /**
         * 选择的组织行
         * @param {?} event
         * @return {?}
         */
            function (event) {
                console.log(event.node.data);
                this.selectRowChanged.next(event.node.data);
            };
        /**
         * 获取当前TreeTable的绑定数据
         */
        /**
         * 获取当前TreeTable的绑定数据
         * @return {?}
         */
        GroupNavComponent.prototype.getTreeData = /**
         * 获取当前TreeTable的绑定数据
         * @return {?}
         */
            function () {
                return this.treeData;
            };
        /**
         * 获取当前TreeTable的组件实例
         */
        /**
         * 获取当前TreeTable的组件实例
         * @return {?}
         */
        GroupNavComponent.prototype.getTreeTableComp = /**
         * 获取当前TreeTable的组件实例
         * @return {?}
         */
            function () {
                return this.farrisTt;
            };
        /**
         * 选中某个节点
         */
        /**
         * 选中某个节点
         * @param {?} id
         * @param {?=} isCheck
         * @return {?}
         */
        GroupNavComponent.prototype.selectNode = /**
         * 选中某个节点
         * @param {?} id
         * @param {?=} isCheck
         * @return {?}
         */
            function (id, isCheck) {
                this.farrisTt.selectNode(id, isCheck);
            };
        /**
         * 清除所有选中值
         */
        /**
         * 清除所有选中值
         * @return {?}
         */
        GroupNavComponent.prototype.unCheckAllNodes = /**
         * 清除所有选中值
         * @return {?}
         */
            function () {
                this.farrisTt.unCheckAllNodes();
            };
        /**
         * 清除所有选中值
         */
        /**
         * 清除所有选中值
         * @param {?} key
         * @return {?}
         */
        GroupNavComponent.prototype.scrollToCurrentNode = /**
         * 清除所有选中值
         * @param {?} key
         * @return {?}
         */
            function (key) {
                if (key) {
                    this.farrisTt.selectNode(key);
                    this.expandParent(key);
                    this.farrisTt.scrollToCurrentNode();
                }
            };
        /** 逐一找到需要展开的父节点，全部展开 */
        /**
         * 逐一找到需要展开的父节点，全部展开
         * @param {?} key
         * @return {?}
         */
        GroupNavComponent.prototype.expandParent = /**
         * 逐一找到需要展开的父节点，全部展开
         * @param {?} key
         * @return {?}
         */
            function (key) {
                /** @type {?} */
                var parent = this.farrisTt.findRowNode(key);
                /** @type {?} */
                var parentNode = parent.parent;
                if (parentNode) {
                    while (parentNode) {
                        this.farrisTt.expandNode(parentNode.id);
                        parentNode = parentNode.parent;
                    }
                }
            };
        GroupNavComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'lib-group-nav',
                        template: "<div #farrisTreeParent class=\"f-utils-fill-flex-column\" style=\"background-color: white\">\r\n    <farris-treetable #farrisTt \r\n    [data]=\"this.treeData\" \r\n    [columns]=\"cols\" \r\n    [idField]=\"'id'\" \r\n    [fixedHeader]=\"fixedHeader\" \r\n    [showHeader]=\"showHeader\" \r\n    [fit]=\"true\" \r\n    [resizableColumns]=\"true\"\r\n    [fitColumns]=\"true\"\r\n    [(showFilterBar)]=\"showFilterBar\"\r\n    (search)=\"searchData($event)\"\r\n    [searchFields]=\"searchFields\"\r\n    [allColumnsTitle]=\"allColumnsTitle\"\r\n    (clearSearchValue)=\"clearSearchValue()\"\r\n    (nodeSelected)=\"onNodeSelect($event)\">\r\n    </farris-treetable>\r\n</div>\r\n",
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        GroupNavComponent.ctorParameters = function () {
            return [
                { type: GroupService },
                { type: SysLocaleService },
                { type: uiNotify.NotifyService },
                { type: uiMessager.MessagerService }
            ];
        };
        GroupNavComponent.propDecorators = {
            farrisTreeParent: [{ type: i0.ViewChild, args: ['farrisTreeParent',] }],
            farrisTt: [{ type: i0.ViewChild, args: ['farrisTt',] }],
            title: [{ type: i0.Input }],
            showHeader: [{ type: i0.Input }],
            fixedHeader: [{ type: i0.Input }],
            withCodeColumn: [{ type: i0.Input }],
            allRoot: [{ type: i0.Input }],
            groupType: [{ type: i0.Input }],
            filter: [{ type: i0.Input }],
            selectKeys: [{ type: i0.Input }],
            needRemote: [{ type: i0.Input }],
            showNodeTypeColumn: [{ type: i0.Input }],
            treeData: [{ type: i0.Input }],
            showFilterBar: [{ type: i0.Input }],
            search: [{ type: i0.Output }],
            selectRowChanged: [{ type: i0.Output }],
            winResize: [{ type: i0.HostListener, args: ['window:resize', [],] }]
        };
        return GroupNavComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var userHelp = 'user';
    /** @type {?} */
    var posHelp = 'position';
    /** @type {?} */
    var roleHelp = 'role';
    /** @type {?} */
    var userGroupHelp = 'userGroup';
    /** @type {?} */
    var posGroupHelp = 'posGroup';
    /** @type {?} */
    var roleGroupHelp = 'roleGroupHelp';
    /** @type {?} */
    var sysOrgHelp = 'sysOrg';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 帮助服务
     */
    var SysFarrisLookupService = /** @class */ (function () {
        function SysFarrisLookupService(http, sysUtil, localService) {
            this.http = http;
            this.sysUtil = sysUtil;
            this.localService = localService;
        }
        /** 初始过滤条件 */
        /**
         * 初始过滤条件
         * @private
         * @param {?} filter
         * @return {?}
         */
        SysFarrisLookupService.prototype.setFilterData = /**
         * 初始过滤条件
         * @private
         * @param {?} filter
         * @return {?}
         */
            function (filter) {
                // 有过滤条件
                if (filter) {
                    /** @type {?} */
                    var str = JSON.stringify(filter);
                    if (str !== '{}') {
                        // 一定要复制
                        this.filterData = JSON.parse(str);
                    }
                    else {
                        this.filterData = null;
                    }
                }
                else {
                    this.filterData = null;
                }
            };
        /**
         * 添加数据
         */
        /**
         * 添加数据
         * @private
         * @param {?} src
         * @param {?} data
         * @return {?}
         */
        SysFarrisLookupService.prototype.setListData = /**
         * 添加数据
         * @private
         * @param {?} src
         * @param {?} data
         * @return {?}
         */
            function (src, data) {
                if (src) {
                    src.push.apply(src, data);
                }
            };
        /** 获取过滤数据 */
        /**
         * 获取过滤数据
         * @private
         * @return {?}
         */
        SysFarrisLookupService.prototype.getFilterData = /**
         * 获取过滤数据
         * @private
         * @return {?}
         */
            function () {
                return this.filterData;
            };
        /**
         * 异步获取版本
         */
        /**
         * 异步获取版本
         * @param {?} url
         * @param {?=} params
         * @return {?}
         */
        SysFarrisLookupService.prototype.getTreeDataAsync = /**
         * 异步获取版本
         * @param {?} url
         * @param {?=} params
         * @return {?}
         */
            function (url, params) {
                var _this = this;
                this.i18nRes = this.localService.getComponentOpt('help');
                // 解析查询json对象
                /** @type {?} */
                var treeParam = JSON.parse(params.searchValue);
                /** @type {?} */
                var farrisSelectedData = params['customData'].farrisSelectedData;
                /** @type {?} */
                var filter = {};
                // filter.withPermission = true;
                /** @type {?} */
                var category = treeParam.category;
                // 1.首次加载
                if (category === 'all') {
                    filter.layer = 1;
                }
                // 2.异步展开下级
                if (category === 'children') {
                    if (treeParam.hasOwnProperty('parentPath')) {
                        filter.path = treeParam.parentPath;
                    }
                    if (treeParam.hasOwnProperty('parentLayer')) {
                        // 后端搜索以下级节点为级数的节点
                        filter.layer = treeParam.parentLayer + 1;
                    }
                    if (treeParam.hasOwnProperty('parentId')) {
                        filter.parentId = treeParam.parentId;
                    }
                }
                // 3.搜索
                if (category === 'search') {
                    /** @type {?} */
                    var searchField = treeParam.searchField;
                    /** @type {?} */
                    var searchValue = treeParam.searchValue.trim();
                    if (searchValue) {
                        filter.isSearchTree = true;
                        switch (searchField) {
                            case 'code':
                                filter.code = searchValue;
                                break;
                            case 'name':
                                filter.name = searchValue;
                                break;
                            // 全部列搜索，之前是'*'
                            case '*':
                            default:
                                filter.codeOrName = searchValue;
                                break;
                        }
                    }
                    else {
                        // 搜索为空，重新加载，默认从1级开始
                        filter.layer = 1;
                    }
                }
                //如果是组织，请求选中数据时
                if (url === '/api/runtime/sys/v1.0/sysOrgs' && category === 'fav') {
                    /** @type {?} */
                    var newUrl = url + '/nullorg';
                    /** @type {?} */
                    var newOpt = {
                        headers: this.sysUtil.getHeader()['headers'],
                        params: { useView: 'true' }
                    };
                    return this.http.get(newUrl, newOpt).pipe(operators.map(( /**
                     * @param {?} val
                     * @return {?}
                     */function (val) {
                        /** @type {?} */
                        var gridResult = {};
                        return gridResult;
                    })));
                }
                // 说明有过滤条件
                if (this.getFilterData()) {
                    filter = Object.assign(this.getFilterData(), filter);
                    i2.SysDataUtil.deleteEmpty(filter);
                }
                /** @type {?} */
                var paramStr = JSON.stringify(filter);
                /** @type {?} */
                var opt = {
                    headers: this.sysUtil.getHeader()['headers'],
                    params: { param: paramStr }
                };
                return this.http.get(url, opt).pipe(operators.map(( /**
                 * @param {?} val
                 * @return {?}
                 */function (val) {
                    /** @type {?} */
                    var gridResult = {};
                    gridResult.selectedData = farrisSelectedData;
                    gridResult.items = [];
                    gridResult.treeInfo = {
                        loadDataType: 'async',
                        layerType: 'parentId',
                        dataField: 'treeInfo',
                        pathField: 'path',
                        layerField: 'layer',
                        parentField: 'parentId',
                        isDetailField: 'isDetail'
                    };
                    gridResult.treeInfo['treeDataIsInit'] = true;
                    gridResult.columns = [
                        { field: 'code', title: _this.i18nRes.codeField },
                        { field: 'name', title: _this.i18nRes.nameField }
                    ];
                    gridResult.searchFields = [
                        { label: _this.i18nRes.codeField, value: 'code' },
                        { label: _this.i18nRes.nameField, value: 'name' }
                    ];
                    // 必须要赋值
                    rxjs.from(val).subscribe(( /**
                     * @param {?} item
                     * @return {?}
                     */function (item) {
                        if (item.layer === 1 || !item.parentId) {
                            item.parentId = null;
                        }
                        item.treeInfo = {
                            parentId: item.parentId,
                            path: item.path,
                            layer: item.layer,
                            isDetail: item.isDetail
                        };
                    }));
                    // 如果是搜索的话，需要自己构造树
                    if (filter.isSearchTree) {
                        gridResult.items = _this.makeTreeByParent(val, 'parentId', 'id');
                        console.log(gridResult.items);
                    }
                    else {
                        // 有结果，普通分层加载
                        if (val && val.length > 0) {
                            rxjs.from(val).subscribe(( /**
                             * @param {?} item
                             * @return {?}
                             */function (item) {
                                gridResult.items.push(_this.convertToTreeNode(item));
                            }));
                        }
                    }
                    return gridResult;
                })));
            };
        /**
         * 将树节点转换为Farris TreeNode
         */
        /**
         * 将树节点转换为Farris TreeNode
         * @private
         * @param {?} src
         * @return {?}
         */
        SysFarrisLookupService.prototype.convertToTreeNode = /**
         * 将树节点转换为Farris TreeNode
         * @private
         * @param {?} src
         * @return {?}
         */
            function (src) {
                /** @type {?} */
                var node = {};
                node.data = src;
                node.children = [];
                node.selectable = src.farris_selectable;
                node.leaf = src.isDetail;
                node.parent = src.parentId;
                // 这一条必须得加，否则不行
                node.id = node.data.id;
                // 搜索的时候最末级添加了这个属性，不需要展开
                if (src.hasOwnProperty('expanded')) {
                    node.expanded = src.expanded;
                }
                return node;
            };
        /** 按照parent码构造树 */
        /**
         * 按照parent码构造树
         * @param {?} datas
         * @param {?} parentField
         * @param {?} keyField
         * @return {?}
         */
        SysFarrisLookupService.prototype.makeTreeByParent = /**
         * 按照parent码构造树
         * @param {?} datas
         * @param {?} parentField
         * @param {?} keyField
         * @return {?}
         */
            function (datas, parentField, keyField) {
                var _this = this;
                // 上級爲空，或者沒有上級的
                /** @type {?} */
                var pathLayerOnes = datas.filter(( /**
                 * @param {?} x
                 * @return {?}
                 */function (x) {
                    if (!x[parentField]) {
                        return true;
                    }
                    // 断层的这种，应该也属于一级
                    /** @type {?} */
                    var p = datas.find(( /**
                     * @param {?} parent
                     * @return {?}
                     */function (parent) { return parent[keyField] === x[parentField]; }));
                    if (!p) {
                        return true;
                    }
                }));
                /** @type {?} */
                var treedata = [];
                pathLayerOnes.forEach(( /**
                 * @param {?} x
                 * @return {?}
                 */function (x) {
                    /** @type {?} */
                    var node = _this.convertToTreeNode(x);
                    treedata.push(node);
                }));
                treedata.forEach(( /**
                 * @param {?} item
                 * @return {?}
                 */function (item) {
                    _this.findChildrenByParent(item, datas, parentField, keyField);
                }));
                return treedata;
            };
        /** 找到子节点 */
        /**
         * 找到子节点
         * @param {?} parent
         * @param {?} datas
         * @param {?} parentField
         * @param {?} keyField
         * @return {?}
         */
        SysFarrisLookupService.prototype.findChildrenByParent = /**
         * 找到子节点
         * @param {?} parent
         * @param {?} datas
         * @param {?} parentField
         * @param {?} keyField
         * @return {?}
         */
            function (parent, datas, parentField, keyField) {
                var _this = this;
                /** @type {?} */
                var childList = datas.filter(( /**
                 * @param {?} childitem
                 * @return {?}
                 */function (childitem) {
                    if (childitem[parentField] === parent.data[keyField]) {
                        return true;
                    }
                }));
                if (childList && childList.length > 0) {
                    childList.forEach(( /**
                     * @param {?} child
                     * @return {?}
                     */function (child) {
                        /** @type {?} */
                        var node = _this.convertToTreeNode(child);
                        parent.children.push(node);
                        _this.findChildrenByParent(node, datas, parentField, keyField);
                    }));
                }
            };
        /**
         * 全部获取版本版本
         */
        /**
         * 全部获取版本版本
         * @param {?} url
         * @param {?=} params
         * @return {?}
         */
        SysFarrisLookupService.prototype.getTreeData = /**
         * 全部获取版本版本
         * @param {?} url
         * @param {?=} params
         * @return {?}
         */
            function (url, params) {
                /** @type {?} */
                var filter = {};
                // 说明有过滤条件
                if (this.getFilterData()) {
                    filter = Object.assign(this.getFilterData(), filter);
                    i2.SysDataUtil.deleteEmpty(filter);
                }
                /** @type {?} */
                var paramStr = JSON.stringify(filter);
                /** @type {?} */
                var opt = {
                    headers: this.sysUtil.getHeader()['headers'],
                    params: { param: paramStr }
                };
                return this.http.get(url, opt).pipe(operators.map(( /**
                 * @param {?} val
                 * @return {?}
                 */function (val) {
                    if (val && val.length > 0) {
                        val.forEach(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) {
                            if (item.layer == 1 || item.parentId == null) {
                                item.parentId = '';
                            }
                            // 满足前端结构
                            item.treeInfo = {
                                parentId: item.parentId,
                                path: item.path,
                                layer: item.layer,
                                isDetail: item.isDetail
                            };
                            item.selectable = false;
                        }));
                    }
                    else {
                        val = [];
                    }
                    return { items: val };
                })));
            };
        /** 获取列表数据 */
        /**
         * 获取列表数据
         * @param {?} url
         * @param {?=} params
         * @return {?}
         */
        SysFarrisLookupService.prototype.getGridData = /**
         * 获取列表数据
         * @param {?} url
         * @param {?=} params
         * @return {?}
         */
            function (url, params) {
                var _this = this;
                /** @type {?} */
                var index = params.pageIndex;
                /** @type {?} */
                var size = params.pageSize;
                /** @type {?} */
                var farrisSelectedData = params['customData'].farrisSelectedData;
                // 目前这个地方有问题，第一次的时候传递不过来
                if (!index) {
                    /** @type {?} */
                    var customData = params['customData'];
                    index = customData.filter.pageIndex;
                    size = customData.filter.pageSize;
                }
                /** @type {?} */
                var param = {
                    pageIndex: index,
                    pageSize: size,
                };
                // 是否有搜索
                /** @type {?} */
                var filterStr = params.searchValue;
                /** @type {?} */
                var filterObj = JSON.parse(filterStr);
                /** @type {?} */
                var searchField = filterObj.searchField;
                /** @type {?} */
                var searchValue = filterObj.searchValue;
                // 有过滤查询,有可能搜索值为空，这时候是不需要搜索的
                if (searchField && searchValue) {
                    switch (searchField) {
                        case 'code':
                            param.code = searchValue;
                            break;
                        case 'name':
                            param.name = searchValue;
                            break;
                        // 全部列搜索，之前是'*'
                        case '*':
                        default:
                            param.codeOrName = searchValue;
                            break;
                    }
                }
                // 说明有过滤条件
                if (this.getFilterData()) {
                    param = Object.assign(this.getFilterData(), param);
                    i2.SysDataUtil.deleteEmpty(param);
                }
                console.log(param);
                /** @type {?} */
                var paramStr = JSON.stringify(param);
                /** @type {?} */
                var opt = {
                    headers: this.sysUtil.getHeader()['headers'],
                    params: { param: paramStr }
                };
                return this.http.get(url, opt).pipe(operators.map(( /**
                 * @param {?} val
                 * @return {?}
                 */function (val) {
                    if (val && val.data) {
                        // 将本次分页的数据附加到当前帮助的列表中
                        _this.setListData(params['customData'].allData, val.data);
                    }
                    return {
                        items: val.data,
                        total: val.totalCount,
                        pageInfo: {
                            pageIndex: index,
                            pageSize: size,
                            pageList: [20, 40, 60, 80, 100],
                            enablePager: true
                        },
                        selectedData: farrisSelectedData
                    };
                })));
            };
        /** 取数 */
        /**
         * 取数
         * @param {?} url
         * @param {?=} params
         * @return {?}
         */
        SysFarrisLookupService.prototype.getData = /**
         * 取数
         * @param {?} url
         * @param {?=} params
         * @return {?}
         */
            function (url, params) {
                console.log(params);
                console.log(url);
                /** @type {?} */
                var filter = params['customData'].filter;
                this.setFilterData(filter);
                /** @type {?} */
                var typeIndex = url.indexOf('\/');
                /** @type {?} */
                var httpUrl = url.substring(typeIndex, url.length);
                /** @type {?} */
                var helpType = url.substring(0, typeIndex);
                switch (helpType) {
                    case userHelp:
                    case posHelp:
                    case roleHelp:
                        return this.getGridData(httpUrl, params);
                    case userGroupHelp:
                    case posGroupHelp:
                    case roleGroupHelp:
                    case sysOrgHelp:
                        return this.getTreeDataAsync(httpUrl, params);
                    default:
                        break;
                }
            };
        SysFarrisLookupService.decorators = [
            { type: i0.Injectable }
        ];
        /** @nocollapse */
        SysFarrisLookupService.ctorParameters = function () {
            return [
                { type: i1.HttpClient },
                { type: i2.SysCommonUtilService },
                { type: SysLocaleService }
            ];
        };
        return SysFarrisLookupService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SysFarrisLookupComponent = /** @class */ (function () {
        function SysFarrisLookupComponent(sysUtil, localService) {
            var _this = this;
            this.sysUtil = sysUtil;
            this.localService = localService;
            /**
             * 宽度
             */
            this.width = 550;
            /**
             * 是否显示用户排序字段，WF使用
             */
            this.showUserOrderField = false;
            /**
             * 显示用户的单位信息，WF使用
             */
            this.showUserUnitInfo = false;
            /**
             * 只读
             */
            this.disabled = false;
            /**
             * 必填
             */
            this.required = true;
            /**
             * 打开帮助后是否有选中数据
             */
            this.enableToSelect = true;
            /**
             * 是否分页，默认true
             */
            this.pagination = true;
            /**
             * 首次加载第几页，默认1
             */
            this.pageIndex = 1;
            /**
             * 每页加载数，默认20
             */
            this.pageSize = 20;
            /**
             * 是否开始数据权限过滤
             */
            this.withPermission = false;
            /**
             * 启用关闭前事件
             */
            this.useBeforeCloseEvent = false;
            this.typeMode = 0;
            /**
             * 帮助选择后事件
             */
            this.afterConfirm = new i0.EventEmitter();
            /**
             * 帮助选择后事件，会返回选中和删除的id
             */
            this.afterConfirmData = new i0.EventEmitter();
            /**
             * 清空后事件-暂时不用
             */
            this.clear = new i0.EventEmitter();
            /**
             * 关闭前-暂时不用
             */
            this.loadSuccessEvent = new i0.EventEmitter();
            this.valueChangedEvent = new i0.EventEmitter();
            this.selectDataEvent = new i0.EventEmitter();
            this.dialogClosed = new i0.EventEmitter();
            this.selectedData = [];
            this.farrisSelectedData = [];
            this.mapFields = { id: 'id', name: 'name' };
            /**
             * 帮助前事件,强行传递过滤条件
             */
            this.beforeOpen = ( /**
             * @param {?} arg
             * @return {?}
             */function (arg) {
                if (!_this.filter) {
                    // 可能调用方不传递，识别为非空
                    _this.filter = {};
                }
                /**
                 * WF使用，如果显示用户的单位信息，需要查询条件中增加如下标识，
                 * 服务器端解析
                 */
                if (_this.showUserUnitInfo) {
                    _this.filter.queryMode = '0002';
                }
                _this.filter.pageIndex = _this.pageIndex;
                _this.filter.pageSize = _this.pageSize;
                _this.allData.length = 0;
                _this.farrisSelectedData = [];
                _this.selectedData = [];
                if (_this.bindingData != null && _this.bindingData.id != null && _this.bindingData.id != "") {
                    _this.selectedData = _this.bindingData.id.split(',');
                    _this.selectedData.forEach(( /**
                     * @param {?} item
                     * @return {?}
                     */function (item) {
                        _this.farrisSelectedData.push({ "id": item });
                    }));
                }
                // 把过滤数据和当前的数据传递到service层
                return rxjs.of({
                    showDialog: true, data: {
                        filter: _this.filter,
                        allData: _this.allData,
                        farrisSelectedData: _this.farrisSelectedData
                    }
                });
            });
            /**
             * 帮助关闭前事件,目前无论是否都会继续走afterOk映射事件
             */
            this.dictPicked = ( /**
             * @param {?} arg
             * @return {?}
             */function (arg) {
                if (_this.useBeforeCloseEvent && _this.beforeClose) {
                    return _this.beforeClose(arg).pipe(operators.map(( /**
                     * @param {?} val
                     * @return {?}
                     */function (val) {
                        // 如果能关闭，调用后事件
                        if (val.closeDialog) {
                            _this.afterConfirm.next(arg);
                        }
                        return val;
                    })));
                }
                else {
                    return rxjs.of({ closeDialog: true });
                }
            });
            /**
             * 帮助映射后事件
             */
            this.afterOk = ( /**
             * @param {?} helpdata
             * @param {?} mapfields
             * @return {?}
             */function (helpdata, mapfields) {
                if (_this.useBeforeCloseEvent) {
                    return;
                }
                if (_this.typeMode && _this.typeMode === 1) {
                    if (!_this.singleSelect) {
                        _this.selectedData = _this.selectedData.filter(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) {
                            if (_this.allData.findIndex(( /**
                             * @param {?} element
                             * @return {?}
                             */function (element) {
                                if (item.hasOwnProperty("id")) {
                                    return element.id === item.id;
                                }
                                else {
                                    return element.id === item;
                                }
                            })) > -1) {
                                return true;
                            }
                            else {
                                return false;
                            }
                        }));
                        _this.selectedData.forEach(( /**
                         * @param {?} item
                         * @param {?} index
                         * @return {?}
                         */function (item, index) {
                            if (!item.hasOwnProperty("id")) {
                                _this.allData.forEach(( /**
                                 * @param {?} element
                                 * @return {?}
                                 */function (element) {
                                    if (element.id === item) {
                                        _this.selectedData[index] = element;
                                    }
                                }));
                            }
                        }));
                        helpdata = _this.selectedData;
                    }
                }
                if (helpdata) {
                    // 多选的话需要处理,帮助已逗号为间隔
                    if (!_this.singleSelect) {
                        /** @type {?} */
                        var idList = helpdata.map(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) { return item.id; }));
                        /** @type {?} */
                        var nameList = helpdata.map(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) { return item.name; }));
                        /** @type {?} */
                        var nameStr = nameList.join();
                        /** @type {?} */
                        var idStr = idList.join();
                        _this.displayTxt = nameStr;
                        _this.bindingData.id = idStr;
                        _this.bindingData.name = nameStr;
                    }
                    else {
                        _this.displayTxt = helpdata.name;
                    }
                }
                // 触发事件
                _this.afterConfirm.next(helpdata);
                // this.afterConfirmData.next(this.selctedData);
            });
            this.onOkClick = ( /**
             * @return {?}
             */function () {
                if (_this.useBeforeCloseEvent) {
                    return;
                }
                /** @type {?} */
                var helpdata;
                if (!_this.selectedData || _this.selectedData.length === 0) {
                    if (_this.singleSelect) {
                        _this.afterConfirm.next(helpdata);
                    }
                    else {
                        _this.afterConfirm.next([]);
                    }
                    _this.farrisLookUp.closeDialog();
                    return;
                }
                _this.selectedData = _this.selectedData.filter(( /**
                 * @param {?} item
                 * @return {?}
                 */function (item) {
                    if (_this.allData.findIndex(( /**
                     * @param {?} element
                     * @return {?}
                     */function (element) {
                        if (item.hasOwnProperty("id")) {
                            return element.id === item.id;
                        }
                        else {
                            return element.id === item;
                        }
                    })) > -1) {
                        return true;
                    }
                    else {
                        return false;
                    }
                }));
                _this.selectedData.forEach(( /**
                 * @param {?} item
                 * @param {?} index
                 * @return {?}
                 */function (item, index) {
                    if (!item.hasOwnProperty("id")) {
                        _this.allData.forEach(( /**
                         * @param {?} element
                         * @return {?}
                         */function (element) {
                            if (element.id === item) {
                                _this.selectedData[index] = element;
                            }
                        }));
                    }
                }));
                helpdata = _this.selectedData;
                if (helpdata && helpdata.length > 0) {
                    // 多选的话需要处理,帮助已逗号为间隔
                    if (!_this.singleSelect) {
                        /** @type {?} */
                        var idList = helpdata.map(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) { return item.id; }));
                        /** @type {?} */
                        var nameList = helpdata.map(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) { return item.name; }));
                        /** @type {?} */
                        var nameStr = nameList.join();
                        /** @type {?} */
                        var idStr = idList.join();
                        _this.displayTxt = nameStr;
                        _this.bindingData.id = idStr;
                        _this.bindingData.name = nameStr;
                        _this.afterConfirm.next(helpdata);
                    }
                    else {
                        /** @type {?} */
                        var nameStr = helpdata[0].name;
                        _this.displayTxt = nameStr;
                        _this.afterConfirm.next(helpdata[0]);
                    }
                }
                else { // 目前清空会走这里，而且两次
                    _this.afterConfirm.next(helpdata);
                }
                // 触发事件
                _this.farrisLookUp.closeDialog();
            });
            this.selectedData = [];
        }
        /**
         * @return {?}
         */
        SysFarrisLookupComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                if (this.localService) {
                    this.i18nRes = this.localService.getComponentOpt('help');
                }
                else {
                    this.i18nRes = HELP_LOCALE_ZHCHS;
                }
                this.columns = [
                    { title: this.i18nRes.codeField, field: 'code', searchField: 'code', width: '160' },
                    { title: this.i18nRes.nameField, field: 'name', searchField: 'name', width: '200' },
                ];
                // 显示用户的单位路径信息，WF使用
                if (this.showUserUnitInfo) {
                    this.width = 850;
                    this.columns.push({ title: this.i18nRes.orgNamePathField, field: 'orgNamePath' });
                    this.width = 850;
                    this.columns.push({ title: this.i18nRes.orgName, field: 'sysOrgName' });
                }
                // 显示排序字段，WF使用
                if (this.showUserOrderField) {
                    this.columns.push({ title: this.i18nRes.orderField, field: 'orderNum' });
                }
                // 初始化
                this.allData = [];
                // webapi初始
                switch (this.sourceType) {
                    case 'position':
                        this.dataUri = posHelp + this.sysUtil.positionWebApi;
                        if (!this.title) {
                            this.title = this.i18nRes.positionTitle;
                        }
                        break;
                    case 'role':
                        this.dataUri = roleHelp + this.sysUtil.roleWebApi;
                        if (!this.title) {
                            this.title = this.i18nRes.roleTitle;
                        }
                        break;
                    case 'user':
                    default:
                        this.dataUri = userHelp + this.sysUtil.userWebApi;
                        if (!this.title) {
                            this.title = this.i18nRes.userTitle;
                        }
                        break;
                }
                // 必须要有，否则初始选中不起作用
                this.farrisLookUp.mapFields = { id: 'id', name: 'name' };
                // 帮助关闭后事件
                this.farrisLookUp.dialogClosed.subscribe(( /**
                 * @param {?} val
                 * @return {?}
                 */function (val) {
                    console.log(val);
                    _this.dialogClosed.next(val);
                }));
            };
        /** 清空后 */
        /**
         * 清空后
         * @param {?} arg
         * @return {?}
         */
        SysFarrisLookupComponent.prototype.afterClear = /**
         * 清空后
         * @param {?} arg
         * @return {?}
         */
            function (arg) {
                this.clear.next(arg);
            };
        /** 手动显示帮助 */
        /**
         * 手动显示帮助
         * @return {?}
         */
        SysFarrisLookupComponent.prototype.showHelp = /**
         * 手动显示帮助
         * @return {?}
         */
            function () {
                this.farrisLookUp.showDialog();
            };
        /**
         * @param {?} arg
         * @return {?}
         */
        SysFarrisLookupComponent.prototype.oncheckedChange = /**
         * @param {?} arg
         * @return {?}
         */
            function (arg) {
                var _this = this;
                if (this.singleSelect) {
                    this.selectedData = [];
                }
                if (arg.isCheck) {
                    arg.data.forEach(( /**
                     * @param {?} item
                     * @return {?}
                     */function (item) {
                        if (_this.selectedData.indexOf(item.id) === -1) {
                            _this.selectedData.push(item);
                        }
                    }));
                }
                else {
                    this.selectedData = this.selectedData.filter(( /**
                     * @param {?} item
                     * @return {?}
                     */function (item) {
                        if (arg.data.findIndex(( /**
                         * @param {?} element
                         * @return {?}
                         */function (element) { return item.hasOwnProperty("id") ? item.id === element.id : item === element.id; })) > -1) {
                            return false;
                        }
                        else {
                            return true;
                        }
                    }));
                }
            };
        SysFarrisLookupComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sys-farris-grid-lookup',
                        template: "<farris-lookup-grid #farrislookup\r\n    [disabled]=\"disabled\"\r\n    [editable]=\"false\"\r\n    [bindingData]=\"bindingData\"\r\n    [(ngModel)]=\"displayTxt\"\r\n    [columns]=\"columns\"\r\n    [singleSelect]=\"singleSelect\"\r\n    [title]=\"title\"\r\n    [displayType]=\"'List'\"\r\n    [uri]=\"dataUri\"\r\n    [idField]=\"'id'\"\r\n    [pageIndex]=\"pageIndex\"\r\n    [pageSize]=\"pageSize\"\r\n    [enableToSelect]=\"enableToSelect\"\r\n    [pagination]=\"pagination\"\r\n    [valueField]=\"'id'\" \r\n    [textField]=\"'name'\"\r\n    (clear)=\"afterClear($event)\"\r\n    [mapFields]=\"mapFields\"\r\n    [dictPicking]=\"beforeOpen\"\r\n    [dictPicked]=\"beforeClose\"\r\n    [mappingFn]= \"afterOk\"\r\n    [dialogWidth]=\"width\"\r\n    (checkedChange)=\"oncheckedChange($event)\"\r\n\r\n    [required]=\"required\"></farris-lookup-grid>\r\n\r\n    <!--\r\n        \u786E\u5B9A\u7684\u81EA\u5B9A\u4E49\u4E8B\u4EF6\uFF0C\u9700\u8981\u5220\u9664\r\n        [okHandler]=\"onOkClick\" -->\r\n\r\n",
                        providers: [
                            { provide: uiLookup.ServerSideToken, useClass: SysFarrisLookupService }
                        ],
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        SysFarrisLookupComponent.ctorParameters = function () {
            return [
                { type: i2.SysCommonUtilService },
                { type: SysLocaleService }
            ];
        };
        SysFarrisLookupComponent.propDecorators = {
            title: [{ type: i0.Input }],
            width: [{ type: i0.Input }],
            showUserOrderField: [{ type: i0.Input }],
            showUserUnitInfo: [{ type: i0.Input }],
            sourceType: [{ type: i0.Input }],
            disabled: [{ type: i0.Input }],
            required: [{ type: i0.Input }],
            enableToSelect: [{ type: i0.Input }],
            singleSelect: [{ type: i0.Input }],
            pagination: [{ type: i0.Input }],
            pageIndex: [{ type: i0.Input }],
            pageSize: [{ type: i0.Input }],
            displayTxt: [{ type: i0.Input }],
            bindingData: [{ type: i0.Input }],
            filter: [{ type: i0.Input }],
            withPermission: [{ type: i0.Input }],
            useBeforeCloseEvent: [{ type: i0.Input }],
            beforeClose: [{ type: i0.Input }],
            typeMode: [{ type: i0.Input }],
            afterConfirm: [{ type: i0.Output }],
            afterConfirmData: [{ type: i0.Output }],
            clear: [{ type: i0.Output }],
            loadSuccessEvent: [{ type: i0.Output }],
            valueChangedEvent: [{ type: i0.Output }],
            selectDataEvent: [{ type: i0.Output }],
            dialogClosed: [{ type: i0.Output }],
            farrisLookUp: [{ type: i0.ViewChild, args: ['farrislookup',] }]
        };
        return SysFarrisLookupComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var HelpFilterMap = /** @class */ (function () {
        function HelpFilterMap() {
            /**
             * 帮助元数据的过滤条件
             */
            this.helpMeta = new Map();
        }
        return HelpFilterMap;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SysFarrisTreeHelpComponent = /** @class */ (function () {
        function SysFarrisTreeHelpComponent(sysUtil, localService, sysHelpService) {
            var _this = this;
            this.sysUtil = sysUtil;
            this.localService = localService;
            this.sysHelpService = sysHelpService;
            /**
             * 是否只读
             */
            this.disabled = false;
            /**
             * 是否分层加载组织
             */
            this.async = false;
            /**
             * 是否单选
             */
            this.singleSelect = true;
            /**
             * 打开帮助后是否有选中数据
             */
            this.enableToSelect = true;
            /**
             * 启用关闭前事件
             */
            this.useBeforeCloseEvent = false;
            /**
             * 是否开始数据权限过滤
             */
            this.withPermission = false;
            /**
             * 帮助选择后事件
             */
            this.afterConfirm = new i0.EventEmitter();
            /**
             * 清空后事件-暂时不用
             */
            this.clear = new i0.EventEmitter();
            this.showAllPathOrg = false;
            this.mapFields = { id: 'id', name: 'name' };
            this.allData = [];
            /**
             * 帮助前事件,强行传递过滤条件
             */
            this.beforeOpen = ( /**
             * @param {?} arg
             * @return {?}
             */function (arg) {
                //如果有过滤条件，附加到url路径
                /** @type {?} */
                var farrisSelectedData = [];
                /** @type {?} */
                var selectedData = [];
                /** @type {?} */
                var temp = 0;
                if (_this.bindingData != null && _this.bindingData.id != null && _this.bindingData.id != "") {
                    selectedData = _this.bindingData.id.split(',');
                    /** @type {?} */
                    var selectedDataName_1 = _this.bindingData.name.split(',');
                    selectedData.forEach(( /**
                     * @param {?} item
                     * @return {?}
                     */function (item) {
                        farrisSelectedData.push({ "id": item, "name": selectedDataName_1[temp] });
                        temp = temp + 1;
                    }));
                }
                return rxjs.of({
                    showDialog: true,
                    data: {
                        filter: _this.filter,
                        allData: _this.allData,
                        farrisSelectedData: farrisSelectedData
                    }
                });
            });
            /**
             * 帮助关闭前事件,目前无论是否都会继续走afterOk映射事件
             */
            this.dictPicked = ( /**
             * @param {?} arg
             * @return {?}
             */function (arg) {
                if (_this.useBeforeCloseEvent && _this.beforeClose) {
                    return _this.beforeClose(arg).pipe(operators.map(( /**
                     * @param {?} val
                     * @return {?}
                     */function (val) {
                        // 如果能关闭，调用后事件
                        if (val.closeDialog) {
                            _this.afterConfirm.next(arg);
                        }
                        return val;
                    })));
                }
                else {
                    return rxjs.of({ closeDialog: true });
                }
            });
            /**
             * 帮助映射后事件
             */
            this.afterOk = ( /**
             * @param {?} helpdata
             * @param {?} mapfields
             * @return {?}
             */function (helpdata, mapfields) {
                // 如果启用了关闭前校验事件，不走这里
                if (_this.useBeforeCloseEvent) {
                    return;
                }
                if (helpdata) {
                    // 多选的话需要处理,帮助已逗号为间隔
                    if (!_this.singleSelect) {
                        /** @type {?} */
                        var idList = helpdata.map(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) { return item.id; }));
                        /** @type {?} */
                        var nameList = helpdata.map(( /**
                         * @param {?} item
                         * @return {?}
                         */function (item) { return item.name; }));
                        _this.bindingData.id = idList.join();
                        _this.bindingData.name = nameList.join();
                        _this.displayTxt = _this.bindingData.name;
                        _this.afterConfirm.next(helpdata);
                    }
                    else {
                        // 帮助显示
                        _this.displayTxt = helpdata.name;
                        if (_this.showAllPathOrg) {
                            _this.sysHelpService.getParentOrg(helpdata).subscribe(( /**
                             * @param {?} items
                             * @return {?}
                             */function (items) {
                                if (items && items.length > 0) {
                                    helpdata.allPathOrg = items;
                                }
                                _this.afterConfirm.next(helpdata);
                                console.log(helpdata);
                            }));
                        }
                        else {
                            _this.afterConfirm.next(helpdata);
                        }
                    }
                }
                else { // 目前清空会走这里，而且两次
                    // 触发事件
                    _this.afterConfirm.next(helpdata);
                    console.log(helpdata);
                }
            });
        }
        /**
         * @return {?}
         */
        SysFarrisTreeHelpComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                if (this.localService) {
                    this.i18nRes = this.localService.getComponentOpt('help');
                }
                else {
                    this.i18nRes = HELP_LOCALE_ZHCHS;
                }
                // 默认不分层加载
                this.treeInfo = {
                    loadDataType: 'all',
                    layerType: 'parentId',
                    dataField: 'treeInfo',
                    pathField: 'path',
                    layerField: 'layer',
                    parentField: 'parentId',
                    isDetailField: 'isDetail',
                    treeDataIsInit: false
                };
                this.columns = [
                    { title: this.i18nRes.codeField, field: 'code', searchField: 'code' },
                    { title: this.i18nRes.nameField, field: 'name', searchField: 'name' },
                ];
                // 异步加载的话需要开启这两个开关
                if (this.async) {
                    this.treeInfo.loadDataType = 'async';
                    this.treeInfo.treeDataIsInit = true;
                }
                // webapi初始
                switch (this.sourceType) {
                    case 'org':
                        this.dataUri = sysOrgHelp + this.sysUtil.sysOrgWebApi;
                        if (!this.title) {
                            this.title = this.i18nRes.userOrgTitle;
                        }
                        break;
                    case 'usergroup':
                        this.dataUri = userGroupHelp + this.sysUtil.userGroupWebApi;
                        if (!this.title) {
                            this.title = this.i18nRes.userGroupTitle;
                        }
                        break;
                    case 'positiongroup':
                        this.dataUri = posGroupHelp + this.sysUtil.posGroupWebApi;
                        if (!this.title) {
                            this.title = this.i18nRes.positionGroupTitle;
                        }
                        break;
                    case 'rolegroup':
                    default:
                        this.dataUri = roleGroupHelp + this.sysUtil.roleGroupWebApi;
                        if (!this.title) {
                            this.title = this.i18nRes.roleGroupTitle;
                        }
                        break;
                }
                // 必须要有，否则初始选中不起作用
                this.farrisLookUp.mapFields = { id: 'id', name: 'name' };
            };
        /** 清空后 */
        /**
         * 清空后
         * @param {?} arg
         * @return {?}
         */
        SysFarrisTreeHelpComponent.prototype.afterClear = /**
         * 清空后
         * @param {?} arg
         * @return {?}
         */
            function (arg) {
                this.clear.next(arg);
            };
        /** 手动显示帮助 */
        /**
         * 手动显示帮助
         * @return {?}
         */
        SysFarrisTreeHelpComponent.prototype.showHelp = /**
         * 手动显示帮助
         * @return {?}
         */
            function () {
                this.farrisLookUp.showDialog();
            };
        /**
         * @param {?} $event
         * @return {?}
         */
        SysFarrisTreeHelpComponent.prototype.search = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                console.log($event);
            };
        SysFarrisTreeHelpComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sys-farris-tree-help',
                        template: "<farris-lookup-grid #farristreelookup\r\n    [(ngModel)]=\"displayTxt\"\r\n    [bindingData]=\"bindingData\"\r\n    [title]=\"title\"\r\n    [remoteSearch]=\"true\"\r\n    [editable]=\"false\"\r\n    [disabled]=\"disabled\"\r\n    [displayType]=\"'TREELIST'\"\r\n    [singleSelect]=\"singleSelect\"\r\n    [enableToSelect]=\"enableToSelect\"\r\n    [uri]=\"dataUri\"\r\n    [columns] = \"columns\"\r\n    [idField]=\"'id'\"\r\n    [treeInfo]=\"treeInfo\"\r\n    [textField]= \"'name'\"\r\n    [valueField]=\"'id'\"\r\n    [mappingFn]= \"afterOk\"\r\n    [mapFields]=\"mapFields\"\r\n    [dictPicking]=\"beforeOpen\"\r\n    [dictPicked]=\"dictPicked\"\r\n    (clear)=\"afterClear($event)\"\r\n    (search)=\"search($event)\"\r\n></farris-lookup-grid>",
                        providers: [
                            { provide: uiLookup.ServerSideToken, useClass: SysFarrisLookupService }
                        ],
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        SysFarrisTreeHelpComponent.ctorParameters = function () {
            return [
                { type: i2.SysCommonUtilService },
                { type: SysLocaleService },
                { type: SysHelpService }
            ];
        };
        SysFarrisTreeHelpComponent.propDecorators = {
            disabled: [{ type: i0.Input }],
            title: [{ type: i0.Input }],
            async: [{ type: i0.Input }],
            sourceType: [{ type: i0.Input }],
            singleSelect: [{ type: i0.Input }],
            displayTxt: [{ type: i0.Input }],
            bindingData: [{ type: i0.Input }],
            filter: [{ type: i0.Input }],
            enableToSelect: [{ type: i0.Input }],
            useBeforeCloseEvent: [{ type: i0.Input }],
            beforeClose: [{ type: i0.Input }],
            withPermission: [{ type: i0.Input }],
            afterConfirm: [{ type: i0.Output }],
            clear: [{ type: i0.Output }],
            showAllPathOrg: [{ type: i0.Input }],
            farrisLookUp: [{ type: i0.ViewChild, args: ['farristreelookup',] }]
        };
        return SysFarrisTreeHelpComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SysManagerUIModule = /** @class */ (function () {
        function SysManagerUIModule() {
        }
        SysManagerUIModule.decorators = [
            { type: i0.NgModule, args: [{
                        imports: [
                            common.CommonModule,
                            forms.FormsModule,
                            i1.HttpClientModule,
                            uiLookup.LookupModule,
                            uiModal.ModalModule.forRoot(),
                            uiMessager.MessagerModule.forRoot(),
                            uiDatagrid.DatagridModule,
                            uiTreetable.TreeTableModule,
                            i2.SysCommonModule
                        ],
                        declarations: [
                            GroupNavComponent,
                            SysFarrisLookupComponent,
                            SysFarrisTreeHelpComponent
                        ],
                        exports: [
                            GroupNavComponent,
                            SysFarrisLookupComponent,
                            SysFarrisTreeHelpComponent
                        ],
                        providers: [
                            HelpFilterMap,
                            SysHelpService,
                            SysExceptionService,
                            i2.SysCommonUtilService,
                            GroupService,
                            cafCommon.LanguageService,
                            ngxCookieService.CookieService,
                            SysLocaleService
                        ]
                    },] }
        ];
        return SysManagerUIModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * 数据权限操作
     */
    var AuthOp = /** @class */ (function () {
        function AuthOp() {
        }
        AuthOp.userManage = 'UserManage';
        AuthOp.userView = 'UserView';
        AuthOp.userAssPosition = 'UserAssPosition';
        AuthOp.positionManage = 'PositionManage';
        AuthOp.positionView = 'PositionView';
        AuthOp.positionAssRole = 'PositionAssRole';
        /**
         * 已经废弃
         */
        AuthOp.positionAssUser = 'PositionAssUser';
        AuthOp.roleManage = 'RoleManage';
        /**
         * 已经废弃
         */
        AuthOp.roleAssPosition = 'RoleAssPosition';
        return AuthOp;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.SysFuncOpHelpModule = SysFuncOpHelpModule;
    exports.SysFuncOpHelpComponent = SysFuncOpHelpComponent;
    exports.SysFuncOperationComponent = SysFuncOperationComponent;
    exports.SysManagerUIModule = SysManagerUIModule;
    exports.AuthOp = AuthOp;
    exports.SysFarrisLookupService = SysFarrisLookupService;
    exports.SysHelpService = SysHelpService;
    exports.userHelp = userHelp;
    exports.posHelp = posHelp;
    exports.roleHelp = roleHelp;
    exports.userGroupHelp = userGroupHelp;
    exports.posGroupHelp = posGroupHelp;
    exports.roleGroupHelp = roleGroupHelp;
    exports.sysOrgHelp = sysOrgHelp;
    exports.SysExceptionService = SysExceptionService;
    exports.GroupNavComponent = GroupNavComponent;
    exports.SysFarrisLookupComponent = SysFarrisLookupComponent;
    exports.SysFarrisTreeHelpComponent = SysFarrisTreeHelpComponent;
    exports.SysLocaleService = SysLocaleService;
    exports.HelpFilterMap = HelpFilterMap;
    exports.ɵc = BizObjectComponent;
    exports.ɵd = BizObjectService;
    exports.ɵa = SysFuncOperationService;
    exports.ɵe = SysFarrisFuncOperationComponent;
    exports.ɵb = SysFrontUtil;
    exports.ɵf = GroupService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
