// @ecp-caf/common-structure/esm5/lib/convertor/data-type/map-type-convertor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DataTypeConvertUtil } from '../utils/data-type-convert-util';
import { ExtensionConvertor } from '../extension/extension-convertor';
var MapTypeConvertor = /** @class */ (function () {
    function MapTypeConvertor() {
    }
    /**
     * @param {?} obj
     * @param {?} datatype
     * @param {?} ctxt
     * @return {?}
     */
    MapTypeConvertor.prototype.convertToObject = /**
     * @param {?} obj
     * @param {?} datatype
     * @param {?} ctxt
     * @return {?}
     */
    function (obj, datatype, ctxt) {
        if (!datatype) {
            return;
        }
        /** @type {?} */
        var maptype = (/** @type {?} */ (datatype));
        obj['kind'] = maptype.kind;
        obj['keyType'] = null;
        if (maptype.keyType) {
            /** @type {?} */
            var kind = maptype.keyType.getKind();
            if (!kind) {
                throw new Error('Map类型序列化时KeyType的TypeKind为空！');
            }
            obj['keyType'] = DataTypeConvertUtil.convertToObject(maptype.keyType, ctxt);
        }
        obj['valueType'] = null;
        if (maptype.valueType) {
            /** @type {?} */
            var kind = maptype.valueType.getKind();
            if (!kind) {
                throw new Error('Map类型序列化时ValueType的TypeKind为空！');
            }
            obj['valueType'] = DataTypeConvertUtil.convertToObject(maptype.valueType, ctxt);
        }
        if (maptype.extension) {
            /** @type {?} */
            var extensionConvertor = new ExtensionConvertor();
            extensionConvertor.addToJObj(maptype, obj);
        }
    };
    /**
     * @param {?} dataType
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    MapTypeConvertor.prototype.initFromObject = /**
     * @param {?} dataType
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    function (dataType, obj, ctxt) {
        /** @type {?} */
        var maptype = (/** @type {?} */ (dataType));
        if (!obj) {
            return;
        }
        maptype.kind = obj['kind'];
        if (obj['keyType']) {
            /** @type {?} */
            var kind = obj['keyType']['kind'];
            if (!kind) {
                throw new Error('解析MapType的KeyType属性时数据类型的TypeKind为空！');
            }
            maptype.keyType = DataTypeConvertUtil.initFromObject(obj['keyType'], ctxt);
        }
        if (obj['valueType']) {
            /** @type {?} */
            var kind = obj['valueType']['kind'];
            if (!kind) {
                throw new Error('解析MapType的ValueType属性时数据类型的TypeKind为空！');
            }
            maptype.valueType = DataTypeConvertUtil.initFromObject(obj['valueType'], ctxt);
        }
        if (obj["extension"]) {
            /** @type {?} */
            var extensionConvertor = new ExtensionConvertor();
            extensionConvertor.addToObj(maptype, obj);
        }
    };
    return MapTypeConvertor;
}());
export { MapTypeConvertor };
