// @ecp-caf/common-structure/esm5/lib/convertor/data-type/structured-type-convertor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PropertyConvertor } from '../property/property-convertor';
import { ExtensionConvertor } from '../extension/extension-convertor';
var StructuredTypeConvertor = /** @class */ (function () {
    function StructuredTypeConvertor() {
    }
    /**
     * @param {?} obj
     * @param {?} datatype
     * @param {?} ctxt
     * @return {?}
     */
    StructuredTypeConvertor.prototype.convertToObject = /**
     * @param {?} obj
     * @param {?} datatype
     * @param {?} ctxt
     * @return {?}
     */
    function (obj, datatype, ctxt) {
        /** @type {?} */
        var structure = (/** @type {?} */ (datatype));
        if (!structure) {
            return;
        }
        obj['kind'] = structure.kind;
        obj['id'] = structure.id;
        obj['code'] = structure.code;
        obj['name'] = structure.name;
        obj['description'] = structure.description;
        obj['properties'] = new Array();
        if (structure.properties) {
            /** @type {?} */
            var propertyConvertor = new PropertyConvertor();
            structure.properties.forEach((/**
             * @param {?} property
             * @return {?}
             */
            function (property) {
                /** @type {?} */
                var kind = property.kind;
                /** @type {?} */
                var serializer = ctxt.getPropertySerializerFactory().get(kind);
                /** @type {?} */
                var propertyObj = serializer.serialize(property, ctxt);
                obj['properties'].push(propertyObj);
            }));
        }
        if (structure.extension) {
            /** @type {?} */
            var extensionConvertor = new ExtensionConvertor();
            extensionConvertor.addToJObj(structure, obj);
        }
    };
    /**
     * @param {?} dataType
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    StructuredTypeConvertor.prototype.initFromObject = /**
     * @param {?} dataType
     * @param {?} obj
     * @param {?} ctxt
     * @return {?}
     */
    function (dataType, obj, ctxt) {
        /** @type {?} */
        var structure = (/** @type {?} */ (dataType));
        if (!obj) {
            return;
        }
        structure.kind = obj['kind'];
        structure.id = obj['id'];
        structure.name = obj['name'];
        structure.code = obj['code'];
        structure.description = obj['description'];
        if (obj['properties']) {
            structure.properties = new Array();
            /** @type {?} */
            var propertyConvertor = new PropertyConvertor();
            obj['properties'].forEach((/**
             * @param {?} propertyObj
             * @return {?}
             */
            function (propertyObj) {
                /** @type {?} */
                var propKind = propertyObj["kind"];
                /** @type {?} */
                var serializer = ctxt.getPropertySerializerFactory().get(propKind);
                /** @type {?} */
                var property = serializer.deserialize(propertyObj, ctxt);
                structure.properties.push(property);
            }));
        }
        if (obj["extension"]) {
            /** @type {?} */
            var extensionConvertor = new ExtensionConvertor();
            extensionConvertor.addToObj(structure, obj);
        }
    };
    return StructuredTypeConvertor;
}());
export { StructuredTypeConvertor };
