// @farris/ui-organization-selector/esm5/utils/api.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/api.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { of } from 'rxjs';
var ApiService = /** @class */ (function () {
    function ApiService(http) {
        this.http = http;
    }
    /**
      * 获取行政组织或系统用户
      * @returns observable
      */
    /**
     * 获取行政组织或系统用户
     * @param {?} param
     * @return {?} observable
     */
    ApiService.prototype.getOrganization = /**
     * 获取行政组织或系统用户
     * @param {?} param
     * @return {?} observable
     */
    function (param) {
        /** @type {?} */
        var options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        return this.http.post(this.absUrl + '/org', param, options);
    };
    /**
     * @param {?} absOrgType
     * @return {?}
     */
    ApiService.prototype.getOrgGroups = /**
     * @param {?} absOrgType
     * @return {?}
     */
    function (absOrgType) {
        /** @type {?} */
        var options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        return this.http.get(this.groupApiUrl + '/' + absOrgType, options);
    };
    /**
     * @param {?} groupId
     * @param {?} absOrgType
     * @return {?}
     */
    ApiService.prototype.getGroupMembers = /**
     * @param {?} groupId
     * @param {?} absOrgType
     * @return {?}
     */
    function (groupId, absOrgType) {
        /** @type {?} */
        var options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        if (!groupId) {
            return of(null);
        }
        /** @type {?} */
        var url = this.groupApiUrl + '/' + groupId + '/absOrgType/' + absOrgType;
        return this.http.get(url, options);
    };
    /**
     * @param {?} groupInfo
     * @return {?}
     */
    ApiService.prototype.saveGroup = /**
     * @param {?} groupInfo
     * @return {?}
     */
    function (groupInfo) {
        /** @type {?} */
        var options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json',
            }),
        };
        return this.http.post(this.groupApiUrl, groupInfo, options);
    };
    ApiService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ApiService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return ApiService;
}());
export { ApiService };
if (false) {
    /** @type {?} */
    ApiService.prototype.absUrl;
    /** @type {?} */
    ApiService.prototype.groupApiUrl;
    /**
     * @type {?}
     * @private
     */
    ApiService.prototype.http;
}
