// @qdp/spread/esm5/lib/spread.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { QdpViewComponent } from './component/qdp-view/qdp-view.component';
import { RenderService } from './service/render/render.service';
import { LoadingModule } from '@farris/ui-loading';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { InputGroupModule } from '@farris/ui-input-group';
import { FDropdownDirectiveTypeModule } from '@farris/ui-dropdown';
import { DndListModule } from '@qdp/common';
import { ComboListModule } from '@farris/ui-combo-list';
import { ColumnsSettingComponent } from './component/columns-setting/columns-setting.component';
import { DimensionSettingComponent } from './component/dimension-setting/dimension-setting.component';
var SpreadModule = /** @class */ (function () {
    function SpreadModule() {
    }
    /**
     * @return {?}
     */
    SpreadModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: SpreadModule,
            providers: [
                RenderService
            ]
        };
    };
    SpreadModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        LoadingModule.forRoot({
                            message: 'Loading...'
                        }),
                        DndListModule,
                        FormsModule,
                        ReactiveFormsModule,
                        ComboListModule,
                        InputGroupModule,
                        FDropdownDirectiveTypeModule,
                    ],
                    declarations: [
                        QdpViewComponent,
                        ColumnsSettingComponent,
                        DimensionSettingComponent,
                        DimensionSettingComponent
                    ],
                    exports: [
                        QdpViewComponent,
                        ColumnsSettingComponent,
                    ]
                },] }
    ];
    return SpreadModule;
}());
export { SpreadModule };
