// @farris/ui-search-box/esm5/lib/search-fields/search-field-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, Injector, Input, Output, ViewChild } from '@angular/core';
var SearchFieldPanelComponent = /** @class */ (function () {
    function SearchFieldPanelComponent(injecotr, cd) {
        this.injecotr = injecotr;
        this.cd = cd;
        this.filterChange = new EventEmitter();
        this.remove = new EventEmitter();
        this.conditions = null;
        this.elRef = this.injecotr.get(ElementRef, null);
    }
    /**
     * @return {?}
     */
    SearchFieldPanelComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} $event
     * @return {?}
     */
    SearchFieldPanelComponent.prototype.onSearchValueChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.conditions = $event;
        this.cd.detectChanges();
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SearchFieldPanelComponent.prototype.onSubmitCondition = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.fieldOptions.value = this.conditions.value;
        if (!this.fieldOptions.value.single) {
            this.fieldOptions.value.value = null;
        }
        else {
            delete this.fieldOptions.value.begin;
            delete this.fieldOptions.value.end;
        }
        this.filterChange.emit(this.conditions);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SearchFieldPanelComponent.prototype.onRemoveSearchField = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.remove.emit(this.fieldOptions);
    };
    SearchFieldPanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-search-field-panel',
                    template: "<div class=\"f-search-field-panel f-filter-panel-inner\">\r\n    <div class=\"f-filter-panel-arrow\" style=\"left: 26px; right: auto;\"></div>\r\n    <div class=\"f-filter-panel-content \" style=\"box-shadow:none\">\r\n        <search-filed-editor-date [field]=\"fieldOptions\" *ngIf=\"fieldOptions.control?.controltype === 'datetime'\"\r\n            class=\"d-flex flex-column\" (valueChange)=\"onSearchValueChange($event)\"></search-filed-editor-date>\r\n\r\n        <search-filed-editor-number [field]=\"fieldOptions\" *ngIf=\"fieldOptions.control?.controltype === 'number'\" \r\n            class=\"d-flex flex-column\" (valueChange)=\"onSearchValueChange($event)\"></search-filed-editor-number>\r\n\r\n        <search-filed-editor-enum [field]=\"fieldOptions\" *ngIf=\"fieldOptions.control?.controltype === 'enum'\" \r\n            class=\"d-flex flex-column\" (valueChange)=\"onSearchValueChange($event)\"></search-filed-editor-enum>\r\n\r\n        <search-field-editor-bool [field]=\"fieldOptions\" *ngIf=\"fieldOptions.control?.controltype === 'dropdown'\" \r\n            class=\"d-flex flex-column\" (valueChange)=\"onSearchValueChange($event)\"></search-field-editor-bool>\r\n    </div>\r\n\r\n    <div class=\"f-filter-panel-footer\">\r\n        <div class=\"filer-panel-clear\">\r\n            <button class=\"btn btn-link\" (click)=\"onRemoveSearchField($event)\" [disabled]=\"fieldOptions?.value | canclear\">{{'searchbox.clear' | locale}}</button>\r\n        </div>\r\n        <div class=\"filter-panel-btns\">\r\n            <button class=\"btn btn-secondary close-field-panel\">{{'searchbox.cancel' | locale}}</button>\r\n            <button class=\"filter-panel-submit btn btn-primary\" #btnSubmit [disabled]=\"conditions?.value | canclear\" (click)=\"onSubmitCondition($event)\">{{'searchbox.ok' | locale}}</button>\r\n        </div>\r\n    </div>\r\n</div>"
                }] }
    ];
    /** @nocollapse */
    SearchFieldPanelComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ChangeDetectorRef }
    ]; };
    SearchFieldPanelComponent.propDecorators = {
        fieldOptions: [{ type: Input }],
        btnSubmit: [{ type: ViewChild, args: ['btnSubmit',] }],
        filterChange: [{ type: Output }],
        remove: [{ type: Output }]
    };
    return SearchFieldPanelComponent;
}());
export { SearchFieldPanelComponent };
if (false) {
    /** @type {?} */
    SearchFieldPanelComponent.prototype.fieldOptions;
    /** @type {?} */
    SearchFieldPanelComponent.prototype.btnSubmit;
    /** @type {?} */
    SearchFieldPanelComponent.prototype.filterChange;
    /** @type {?} */
    SearchFieldPanelComponent.prototype.remove;
    /** @type {?} */
    SearchFieldPanelComponent.prototype.conditions;
    /** @type {?} */
    SearchFieldPanelComponent.prototype.elRef;
    /**
     * @type {?}
     * @private
     */
    SearchFieldPanelComponent.prototype.injecotr;
    /**
     * @type {?}
     * @private
     */
    SearchFieldPanelComponent.prototype.cd;
}
