// @gsp-wf/rtdevkit/esm5/lib/entity/complete-response.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 指定参与者信息：需要指定参与者的活动、待指定用户列表
 */
var /**
 * 指定参与者信息：需要指定参与者的活动、待指定用户列表
 */
AssigneeInfo = /** @class */ (function () {
    function AssigneeInfo() {
    }
    return AssigneeInfo;
}());
/**
 * 指定参与者信息：需要指定参与者的活动、待指定用户列表
 */
export { AssigneeInfo };
if (false) {
    /**
     * 工作项ID
     * @type {?}
     */
    AssigneeInfo.prototype.workItemId;
    /**
     * 参与者列表
     * @type {?}
     */
    AssigneeInfo.prototype.assigneeIds;
    /**
     * 参与者用户列表
     * @type {?}
     */
    AssigneeInfo.prototype.assigneeUsers;
    /**
     * 是否允许指派
     * @type {?}
     */
    AssigneeInfo.prototype.selectAssignee;
    /**
     * 是否多选
     * @type {?}
     */
    AssigneeInfo.prototype.multiSelect;
    /**
     * 是否允许全员指派
     * @type {?}
     */
    AssigneeInfo.prototype.allowGlobalAssign;
    /**
     * 指派范围
     * @type {?}
     */
    AssigneeInfo.prototype.organizationScope;
}
/**
 * 加签移交征询范围设置
 */
var /**
 * 加签移交征询范围设置
 */
OrganizationScope = /** @class */ (function () {
    function OrganizationScope() {
    }
    return OrganizationScope;
}());
/**
 * 加签移交征询范围设置
 */
export { OrganizationScope };
if (false) {
    /** @type {?} */
    OrganizationScope.prototype.id;
    /** @type {?} */
    OrganizationScope.prototype.code;
    /** @type {?} */
    OrganizationScope.prototype.name;
    /** @type {?} */
    OrganizationScope.prototype.parentId;
    /** @type {?} */
    OrganizationScope.prototype.path;
    /** @type {?} */
    OrganizationScope.prototype.detail;
    /** @type {?} */
    OrganizationScope.prototype.layer;
    /** @type {?} */
    OrganizationScope.prototype.includeChildren;
    /** @type {?} */
    OrganizationScope.prototype.parentsUntil;
    /** @type {?} */
    OrganizationScope.prototype.principalType;
}
/**
 * 根据流程指定下一步参与者参数
 */
var /**
 * 根据流程指定下一步参与者参数
 */
AssigneeInfoByProcInstIdParameter = /** @class */ (function () {
    function AssigneeInfoByProcInstIdParameter() {
    }
    return AssigneeInfoByProcInstIdParameter;
}());
/**
 * 根据流程指定下一步参与者参数
 */
export { AssigneeInfoByProcInstIdParameter };
if (false) {
    /**
     * 流程实例ID
     * @type {?}
     */
    AssigneeInfoByProcInstIdParameter.prototype.procInstId;
    /**
     * 指定参与者信息
     * @type {?}
     */
    AssigneeInfoByProcInstIdParameter.prototype.assigneeInfos;
}
/**
 * 根据流程指定下一步参与者参数
 */
var /**
 * 根据流程指定下一步参与者参数
 */
AssigneeInfoByWorkItemIdParameter = /** @class */ (function () {
    function AssigneeInfoByWorkItemIdParameter() {
    }
    return AssigneeInfoByWorkItemIdParameter;
}());
/**
 * 根据流程指定下一步参与者参数
 */
export { AssigneeInfoByWorkItemIdParameter };
if (false) {
    /** @type {?} */
    AssigneeInfoByWorkItemIdParameter.prototype.workItemId;
    /** @type {?} */
    AssigneeInfoByWorkItemIdParameter.prototype.assigneeIds;
    /** @type {?} */
    AssigneeInfoByWorkItemIdParameter.prototype.comment;
}
/**
 * 完成工作项返回结果
 */
var /**
 * 完成工作项返回结果
 */
CompleteResponse = /** @class */ (function () {
    function CompleteResponse() {
    }
    return CompleteResponse;
}());
/**
 * 完成工作项返回结果
 */
export { CompleteResponse };
if (false) {
    /** @type {?} */
    CompleteResponse.prototype.dataId;
    /** @type {?} */
    CompleteResponse.prototype.needInteraction;
    /** @type {?} */
    CompleteResponse.prototype.nextAssigneeInfos;
    /** @type {?} */
    CompleteResponse.prototype.procInstId;
    /** @type {?} */
    CompleteResponse.prototype.nextFlowNodes;
    /** @type {?} */
    CompleteResponse.prototype.procDefs;
}
var User = /** @class */ (function () {
    function User() {
    }
    return User;
}());
export { User };
if (false) {
    /** @type {?} */
    User.prototype.id;
    /** @type {?} */
    User.prototype.code;
    /** @type {?} */
    User.prototype.name;
    /** @type {?} */
    User.prototype.note;
    /** @type {?} */
    User.prototype.orgId;
    /** @type {?} */
    User.prototype.orgCode;
    /** @type {?} */
    User.prototype.orgName;
    /** @type {?} */
    User.prototype.orgNamePath;
    /** @type {?} */
    User.prototype.orderNum;
}
var ProcessDefinition = /** @class */ (function () {
    function ProcessDefinition() {
    }
    return ProcessDefinition;
}());
export { ProcessDefinition };
if (false) {
    /** @type {?} */
    ProcessDefinition.prototype.procDefId;
    /** @type {?} */
    ProcessDefinition.prototype.procDefKey;
    /** @type {?} */
    ProcessDefinition.prototype.procDefName;
    /** @type {?} */
    ProcessDefinition.prototype.suspensionState;
    /** @type {?} */
    ProcessDefinition.prototype.version;
    /** @type {?} */
    ProcessDefinition.prototype.description;
}
