// @gsp-wf/wf-process-model/esm2015/lib/model/extension/ExecuteForm.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { ActualParameter } from "../parameter/WFActualParameter";
export class ExecuteForm extends ExtensionElement {
    /**
     * @param {?} model
     * @param {?=} id
     * @param {?=} name
     * @param {?=} actualParameters
     */
    constructor(model, id, name, actualParameters) {
        super(model);
        this.ClrTypeID = ExecuteForm.Clr_Type_ID;
        this.Id = id || '';
        this.Name = name || '';
        this.ActualParameters = actualParameters || [];
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ExecuteForm(this.Model);
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.FormType = null;
        this.Id = null;
        this.Name = null;
        this.ActualParameters = [];
        this.ComponentId = null;
        this.formFormatId = null;
        this.formatConfigId = null;
        this.TerminalType = TerminalType.PC;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "FormType", this.FormType);
        BpmnModelHelper.SetAttributeValue(obj, "Id", this.Id);
        BpmnModelHelper.SetAttributeValue(obj, "Name", this.Name);
        BpmnModelHelper.SetAttributeValue(obj, "ComponentId", this.ComponentId);
        BpmnModelHelper.SetAttributeValue(obj, "TerminalType", this.TerminalType);
        BpmnModelHelper.SetAttributeValue(obj, "FormFormatId", this.formFormatId);
        BpmnModelHelper.SetAttributeValue(obj, "FormatConfigId", this.formatConfigId);
        /** @type {?} */
        let actualParametersElements = [];
        if (this.ActualParameters !== null && this.ActualParameters.length > 0) {
            this.ActualParameters.forEach((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                actualParametersElements.push(item.ConvertToJsonObject());
            }));
        }
        BpmnModelHelper.SetAttributeValue(obj, "ActualParameters", actualParametersElements);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.FormType = obj.FormType;
        this.Id = obj.Id;
        this.Name = obj.Name;
        this.ComponentId = obj.ComponentId;
        this.TerminalType = (/** @type {?} */ (obj.TerminalType));
        this.formFormatId = ExtensionMethod.GetAttributeValue4String(obj, "FormFormatId");
        this.formatConfigId = ExtensionMethod.GetAttributeValue4String(obj, "FormatConfigId");
        /** @type {?} */
        let actualParameters = (/** @type {?} */ (obj.ActualParameters));
        if (actualParameters !== null && actualParameters.length > 0) {
            actualParameters.forEach((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                /** @type {?} */
                let actualParameter = new ActualParameter(this.Model);
                actualParameter.LoadFromJsonObject(item);
                this.ActualParameters.push(actualParameter);
            }));
        }
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const executeForm = new ExecuteForm(this.Model);
        executeForm.Id = BpmnModelHelper.GenerateElementId();
        executeForm.Name = this.Name;
        executeForm.FormType = this.FormType;
        executeForm.formFormatId = this.formFormatId;
        executeForm.formatConfigId = this.formatConfigId;
        executeForm.TerminalType = this.TerminalType;
        executeForm.ComponentId = this.ComponentId;
        executeForm.ActualParameters = new Array();
        if (this.ActualParameters && this.ActualParameters.length > 0) {
            for (const p of this.ActualParameters) {
                /** @type {?} */
                const parameter = p.clone();
                executeForm.ActualParameters.push(parameter);
            }
        }
        return executeForm;
    }
}
ExecuteForm.Clr_Type_ID = "wf.ExecuteForm";
if (false) {
    /** @type {?} */
    ExecuteForm.Clr_Type_ID;
    /** @type {?} */
    ExecuteForm.prototype.ClrTypeID;
    /** @type {?} */
    ExecuteForm.prototype.FormType;
    /** @type {?} */
    ExecuteForm.prototype.Id;
    /** @type {?} */
    ExecuteForm.prototype.Name;
    /** @type {?} */
    ExecuteForm.prototype.ComponentId;
    /** @type {?} */
    ExecuteForm.prototype.TerminalType;
    /** @type {?} */
    ExecuteForm.prototype.ActualParameters;
    /** @type {?} */
    ExecuteForm.prototype.formFormatId;
    /** @type {?} */
    ExecuteForm.prototype.formatConfigId;
}
BpmnModelHelper.RegisterElementType(new ExecuteForm(new BpmnModel()));
/** @enum {string} */
const TerminalType = {
    PC: 'PC',
    MOBILE: 'MOBILE',
};
export { TerminalType };
