// @gsp-wf/wf-process-model/esm2015/lib/model/extension/JumpScope.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement } from "@edp-pmf/bpmn-model";
import { JumpScopeCategory } from "./JumpScopeCategory";
import { JumpScopeDefaultType } from "./JumpScopeDefaultType";
export class JumpScope extends ExtensionElement {
    /**
     * @param {?} model
     * @param {?=} category
     * @param {?=} DefaultType
     */
    constructor(model, category, DefaultType) {
        super(model);
        this.ClrTypeID = JumpScope.Clr_Type_ID;
        this.Category = category || JumpScopeCategory.SpecifiedSet;
        this.DefaultType = DefaultType || JumpScopeDefaultType.FirstNode;
        this.DestinationActivities = new Array();
        this.DefaultActivity = '';
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new JumpScope(this.Model);
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.Category = JumpScopeCategory.Global;
        this.DefaultType = JumpScopeDefaultType.FirstNode;
        this.DestinationActivities = new Array();
        this.DefaultActivity = '';
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "Category", this.Category);
        BpmnModelHelper.SetAttributeValue(obj, "DefaultType", this.DefaultType);
        BpmnModelHelper.SetAttributeValue(obj, "DestinationActivities", this.DestinationActivities);
        BpmnModelHelper.SetAttributeValue(obj, "DefaultActivity", this.DefaultActivity);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.Category = obj.Category;
        this.DefaultType = obj.DefaultType;
        this.DestinationActivities = obj.DestinationActivities;
        this.DefaultActivity = obj.DefaultActivity;
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const jumpScope = new JumpScope(this.Model);
        jumpScope.Category = this.Category;
        jumpScope.DefaultType = this.DefaultType;
        jumpScope.DefaultActivity = this.DefaultActivity;
        jumpScope.DestinationActivities = new Array();
        if (this.DestinationActivities && this.DestinationActivities.length > 0) {
            for (const d of this.DestinationActivities) {
                jumpScope.DestinationActivities.push(d);
            }
        }
        return jumpScope;
    }
}
JumpScope.Clr_Type_ID = "wf.JumpScope";
if (false) {
    /** @type {?} */
    JumpScope.Clr_Type_ID;
    /** @type {?} */
    JumpScope.prototype.ClrTypeID;
    /** @type {?} */
    JumpScope.prototype.Category;
    /** @type {?} */
    JumpScope.prototype.DefaultType;
    /** @type {?} */
    JumpScope.prototype.DestinationActivities;
    /** @type {?} */
    JumpScope.prototype.DefaultActivity;
}
BpmnModelHelper.RegisterElementType(new JumpScope(new BpmnModel()));
