// @gsp-wf/wf-process-model/esm2015/lib/model/extension/OtherAssignActivityNodeInfo.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { Condition } from "./condition-expression/Condition";
export class OtherAssignActivityNodeInfo extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = OtherAssignActivityNodeInfo.Clr_Type_ID;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "ActivityDefinitionId", this.activityDefinitionId);
        BpmnModelHelper.SetAttributeValue(obj, "Required", this.required);
        if (this.condition) {
            BpmnModelHelper.SetAttributeValue(obj, "Condition", this.condition.ConvertToJsonObject());
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.activityDefinitionId = ExtensionMethod.GetAttributeValue4String(obj, "ActivityDefinitionId");
        this.required = ExtensionMethod.GetAttributeValue4Boolean(obj, "Required");
        /** @type {?} */
        const conditionElement = ExtensionMethod.GetAttributeValue(obj, 'Condition');
        if (conditionElement) {
            /** @type {?} */
            const condition = new Condition(this.Model);
            condition.LoadFromJsonObject(conditionElement);
            this.condition = condition;
        }
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const otherAssignActivityNodeInfo = new OtherAssignActivityNodeInfo(this.Model);
        otherAssignActivityNodeInfo.activityDefinitionId = this.activityDefinitionId;
        if (this.condition) {
            otherAssignActivityNodeInfo.condition = this.condition.clone();
        }
        otherAssignActivityNodeInfo.required = this.required;
        return otherAssignActivityNodeInfo;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new OtherAssignActivityNodeInfo(this.Model);
    }
}
OtherAssignActivityNodeInfo.Clr_Type_ID = "wf.OtherAssignActivityNodeInfo";
if (false) {
    /** @type {?} */
    OtherAssignActivityNodeInfo.Clr_Type_ID;
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.ClrTypeID;
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.activityDefinitionId;
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.condition;
    /** @type {?} */
    OtherAssignActivityNodeInfo.prototype.required;
}
BpmnModelHelper.RegisterElementType(new OtherAssignActivityNodeInfo(new BpmnModel()));
