// @gsp-wf/wf-process-model/esm2015/lib/model/extension/OtherAssignNodes.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement, ExtensionMethod } from "@edp-pmf/bpmn-model";
import { OtherAssignActivityNodeInfo } from "./OtherAssignActivityNodeInfo";
export class OtherAssignNodes extends ExtensionElement {
    /**
     * @param {?} model
     */
    constructor(model) {
        super(model);
        this.ClrTypeID = OtherAssignNodes.Clr_Type_ID;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.otherAssignActivityNodeInfos = new Array();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        if (this.otherAssignActivityNodeInfos && this.otherAssignActivityNodeInfos.length > 0) {
            /** @type {?} */
            const destArray = new Array();
            for (const item of this.otherAssignActivityNodeInfos) {
                destArray.push(item.ConvertToJsonObject());
            }
            BpmnModelHelper.SetAttributeValue(obj, "OtherAssignActivityNodeInfos", destArray);
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        /** @type {?} */
        const destArrayElement = ExtensionMethod.GetAttributeValue(obj, "OtherAssignActivities");
        if (destArrayElement && destArrayElement.length > 0) {
            for (const item of destArrayElement) {
                this.otherAssignActivities.push(item);
            }
        }
        /** @type {?} */
        const newDestArrayElement = ExtensionMethod.GetAttributeValue(obj, "OtherAssignActivityNodeInfos");
        if (newDestArrayElement && newDestArrayElement.length > 0) {
            for (const item of newDestArrayElement) {
                /** @type {?} */
                const otherInfo = new OtherAssignActivityNodeInfo(this.Model);
                otherInfo.LoadFromJsonObject(item);
                this.otherAssignActivityNodeInfos.push(otherInfo);
            }
        }
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const assignScope = new OtherAssignNodes(this.Model);
        assignScope.otherAssignActivities = new Array();
        if (this.otherAssignActivities && this.otherAssignActivities.length > 0) {
            for (const d of this.otherAssignActivities) {
                assignScope.otherAssignActivities.push(d);
            }
        }
        assignScope.otherAssignActivityNodeInfos = new Array();
        if (this.otherAssignActivityNodeInfos && this.otherAssignActivityNodeInfos.length > 0) {
            for (const d of this.otherAssignActivityNodeInfos) {
                assignScope.otherAssignActivityNodeInfos.push(d.clone());
            }
        }
        return assignScope;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new OtherAssignNodes(this.Model);
    }
}
OtherAssignNodes.Clr_Type_ID = "wf.OtherAssignNodes";
if (false) {
    /** @type {?} */
    OtherAssignNodes.Clr_Type_ID;
    /** @type {?} */
    OtherAssignNodes.prototype.ClrTypeID;
    /** @type {?} */
    OtherAssignNodes.prototype.otherAssignActivities;
    /** @type {?} */
    OtherAssignNodes.prototype.otherAssignActivityNodeInfos;
}
BpmnModelHelper.RegisterElementType(new OtherAssignNodes(new BpmnModel()));
