// @gsp-wf/wf-process-model/esm2015/lib/model/extension/ProcessHeader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, ExtensionElement } from "@edp-pmf/bpmn-model";
import { StartMode } from "./StartMode";
export class ProcessHeader extends ExtensionElement {
    /**
     * @param {?} model
     * @param {?=} bizDefKey
     * @param {?=} bizObjId
     * @param {?=} bizDefName
     * @param {?=} startMode
     */
    constructor(model, bizDefKey, bizObjId, bizDefName, startMode) {
        super(model);
        this.ClrTypeID = "wf.ProcessHeader";
        this.BizDefKey = bizDefKey || '';
        this.BizObjId = bizObjId || '';
        this.BizDefName = bizDefName || '';
        this.allowPreAssign = false;
        this.enablePriority = false;
        this.enableWholeOrderRetrieve = false;
        this.startMode = startMode || StartMode.DIRECT;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ProcessHeader(this.Model);
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.BizDefKey = null;
        this.BizObjId = null;
        this.BizDefName = null;
        this.allowPreAssign = false;
        this.enablePriority = false;
        this.enableWholeOrderRetrieve = false;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "BizDefKey", this.BizDefKey);
        BpmnModelHelper.SetAttributeValue(obj, "BizObjId", this.BizObjId);
        BpmnModelHelper.SetAttributeValue(obj, "BizDefName", this.BizDefName);
        BpmnModelHelper.SetAttributeValue(obj, "AllowPreAssign", this.allowPreAssign);
        BpmnModelHelper.SetAttributeValue(obj, "EnablePriority", this.enablePriority);
        BpmnModelHelper.SetAttributeValue(obj, "EnableWholeOrderRetrieve", this.enableWholeOrderRetrieve);
        BpmnModelHelper.SetAttributeValue(obj, "StartMode", this.startMode);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.BizDefKey = obj.BizDefKey;
        this.BizObjId = obj.BizObjId;
        this.BizDefName = obj.BizDefName;
        this.allowPreAssign = obj.AllowPreAssign;
        this.enablePriority = obj.EnablePriority;
        this.enableWholeOrderRetrieve = obj.EnableWholeOrderRetrieve;
        if (obj.StartMode) {
            this.startMode = obj.StartMode;
        }
        else if (obj.AccessLevel) {
            switch (obj.AccessLevel) {
                case 'PUBLIC':
                    this.startMode = StartMode.DIRECT;
                    break;
                case 'PROTECTED':
                    this.startMode = StartMode.EXTERNAL_PROCESS;
                    break;
                case 'PRIVATE':
                    this.startMode = StartMode.SUB_PROCESS;
                    break;
            }
        }
    }
}
if (false) {
    /** @type {?} */
    ProcessHeader.prototype.ClrTypeID;
    /** @type {?} */
    ProcessHeader.prototype.BizDefKey;
    /** @type {?} */
    ProcessHeader.prototype.BizDefName;
    /** @type {?} */
    ProcessHeader.prototype.BizObjId;
    /** @type {?} */
    ProcessHeader.prototype.allowPreAssign;
    /** @type {?} */
    ProcessHeader.prototype.enablePriority;
    /** @type {?} */
    ProcessHeader.prototype.enableWholeOrderRetrieve;
    /** @type {?} */
    ProcessHeader.prototype.startMode;
}
BpmnModelHelper.RegisterElementType(new ProcessHeader(new BpmnModel()));
