// @gsp-cmp/coderule-selector/esm5/lib/model/code-rule-base-info.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/model/code-rule-base-info.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CodeRuleBaseInfo = /** @class */ (function () {
    function CodeRuleBaseInfo() {
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    CodeRuleBaseInfo.prototype.init = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        this.id = obj["id"];
        this.code = obj["code"];
        this.name = obj["name"];
        this.serviceUnit = obj["serviceUnit"];
        this.origin = obj["origin"];
    };
    return CodeRuleBaseInfo;
}());
export { CodeRuleBaseInfo };
if (false) {
    /** @type {?} */
    CodeRuleBaseInfo.prototype.id;
    /** @type {?} */
    CodeRuleBaseInfo.prototype.code;
    /** @type {?} */
    CodeRuleBaseInfo.prototype.name;
    /** @type {?} */
    CodeRuleBaseInfo.prototype.serviceUnit;
    /** @type {?} */
    CodeRuleBaseInfo.prototype.origin;
}
