// @edp-aif/runtime-api/fesm5/edp-aif-runtime-api.js
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { JsonConvertableImpl, BpmnModelJsonHelper, PushCreationMode, SlimCreationRuleEntityEx, FormSettings, GeneralTextEntity } from '@edp-aif/common-api';
import { TreeTableModule } from '@farris/ui-treetable';
import { NotifyModule } from '@farris/ui-notify';
import { MessagerModule } from '@farris/ui-messager';
import { NavigationService } from '@farris/command-services';
import { __extends, __values } from 'tslib';
import { InjectionToken, Injectable, Component, Inject, LOCALE_ID, ViewChild, Pipe, NgModule, SystemJsNgModuleLoader, NgModuleFactoryLoader, Injector, ComponentFactoryResolver, ReflectiveInjector } from '@angular/core';
import { ModalModule, BsModalService } from '@farris/ui-modal';
import { Repository, FrameContext } from '@farris/devkit';
import { of, Subject, from } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MappingAssociation = /** @class */ (function (_super) {
    __extends(MappingAssociation, _super);
    function MappingAssociation() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    MappingAssociation.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.relationName = null;
        this.srcRootTableName = null;
        this.srcRootDataId = null;
        this.srcParentTableName = null;
        this.srcParentDataId = null;
        this.srcTableName = null;
        this.srcDataId = null;
        this.tarRootTableName = null;
        this.tarRootDataId = null;
        this.tarParentTableName = null;
        this.tarParentDataId = null;
        this.tarTableName = null;
        this.tarDataId = null;
        this.srcTableNamePath = null;
        this.srcDataIdPath = null;
        this.tarTableNamePath = null;
        this.tarDataIdPath = null;
        this.mappingUnitId = null;
        this.associationType = null;
    };
    /**
     * @return {?}
     */
    MappingAssociation.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "relationName", this.relationName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcRootTableName", this.srcRootTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcRootDataId", this.srcRootDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcParentTableName", this.srcParentTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcParentDataId", this.srcParentDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcTableName", this.srcTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcDataId", this.srcDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarRootTableName", this.tarRootTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarRootDataId", this.tarRootDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarParentTableName", this.tarParentTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarParentDataId", this.tarParentDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarTableName", this.tarTableName);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarDataId", this.tarDataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcTableNamePath", this.srcTableNamePath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "srcDataIdPath", this.srcDataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarTableNamePath", this.tarTableNamePath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "tarDataIdPath", this.tarDataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingUnitId", this.mappingUnitId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "associationType", this.associationType);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    MappingAssociation.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.relationName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "relationName");
        this.srcRootTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcRootTableName");
        this.srcRootDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcRootDataId");
        this.srcParentTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcParentTableName");
        this.srcParentDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcParentDataId");
        this.srcTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcTableName");
        this.srcDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcDataId");
        this.tarRootTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarRootTableName");
        this.tarRootDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarRootDataId");
        this.tarParentTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarParentTableName");
        this.tarParentDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarParentDataId");
        this.tarTableName = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarTableName");
        this.tarDataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarDataId");
        this.srcTableNamePath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcTableNamePath");
        this.srcDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "srcDataIdPath");
        this.tarTableNamePath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarTableNamePath");
        this.tarDataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "tarDataIdPath");
        this.mappingUnitId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "mappingUnitId");
        this.associationType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "associationType");
    };
    return MappingAssociation;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CreationRuleExecutorAssociations = /** @class */ (function (_super) {
    __extends(CreationRuleExecutorAssociations, _super);
    function CreationRuleExecutorAssociations() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isAllAssociation = false;
        _this.associations = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    CreationRuleExecutorAssociations.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.creationRuleId = null;
        this.creationRuleInstanceId = null;
        this.bizFlowchartId = null;
        this.creationProcessId = null;
        this.creationProcessInstanceId = null;
        this.isAllAssociation = false;
        this.associations = new Array();
    };
    /**
     * @return {?}
     */
    CreationRuleExecutorAssociations.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessId", this.creationProcessId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessInstanceId", this.creationProcessInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isAllAssociation", this.isAllAssociation);
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CreationRuleExecutorAssociations.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.creationRuleInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.creationProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessId");
        this.creationProcessInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessInstanceId");
        this.isAllAssociation = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isAllAssociation");
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
    };
    return CreationRuleExecutorAssociations;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BefContentVariable = /** @class */ (function (_super) {
    __extends(BefContentVariable, _super);
    function BefContentVariable() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.associations = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    BefContentVariable.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.upperBeFuncSessionId = null;
        this.associations = new Array();
    };
    /**
     * @return {?}
     */
    BefContentVariable.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upperBeFuncSessionId", this.upperBeFuncSessionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BefContentVariable.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.upperBeFuncSessionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upperBeFuncSessionId");
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", CreationRuleExecutorAssociations);
    };
    return BefContentVariable;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MappingResult = /** @class */ (function (_super) {
    __extends(MappingResult, _super);
    function MappingResult() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    MappingResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.voId = null;
        this.changeDetail = null;
    };
    /**
     * @return {?}
     */
    MappingResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "changeDetail", this.changeDetail);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    MappingResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
        this.changeDetail = JSON.parse(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "changeDetail"));
    };
    return MappingResult;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CreationMappingResult = /** @class */ (function (_super) {
    __extends(CreationMappingResult, _super);
    function CreationMappingResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.associations = new Array();
        _this.entityData = null;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    CreationMappingResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.associations = new Array();
        this.entityData = null;
    };
    /**
     * @return {?}
     */
    CreationMappingResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        BpmnModelJsonHelper.SetAttributeValue(obj, "entityData", this.entityData);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CreationMappingResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
        this.entityData = BpmnModelJsonHelper.GetAttributeValue(obj, "entityData");
    };
    return CreationMappingResult;
}(MappingResult));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CreationRuleExecutorArgs = /** @class */ (function (_super) {
    __extends(CreationRuleExecutorArgs, _super);
    function CreationRuleExecutorArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.sourceVoEntities = new Array();
        _this.targetVoProxyEntities = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    CreationRuleExecutorArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.sourceVoEntities = new Array();
        this.creationRuleId = null;
        this.bizFlowchartId = null;
        this.sourceVoId = null;
        this.targetVoProxyId = null;
        this.targetVoProxyEntities = new Array();
    };
    /**
     * @return {?}
     */
    CreationRuleExecutorArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntities", this.sourceVoEntities);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyId", this.targetVoProxyId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyEntities", this.targetVoProxyEntities);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CreationRuleExecutorArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.sourceVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.sourceVoEntities = BpmnModelJsonHelper.GetAttributeValue(obj, "sourceVoEntities");
        this.targetVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyId");
        this.targetVoProxyEntities = BpmnModelJsonHelper.GetAttributeValue(obj, "targetVoProxyEntities");
    };
    return CreationRuleExecutorArgs;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CreationRuleExecutorResult = /** @class */ (function (_super) {
    __extends(CreationRuleExecutorResult, _super);
    function CreationRuleExecutorResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isAllAssociation = false;
        _this.mappingResults = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    CreationRuleExecutorResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.creationRuleId = null;
        this.creationRuleInstanceId = null;
        this.bizFlowchartId = null;
        this.isAllAssociation = false;
        this.mappingResults = new Array();
    };
    /**
     * @return {?}
     */
    CreationRuleExecutorResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isAllAssociation", this.isAllAssociation);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingResults", this.mappingResults);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    CreationRuleExecutorResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.creationRuleInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.isAllAssociation = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isAllAssociation");
        this.mappingResults = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingResults", CreationMappingResult);
    };
    return CreationRuleExecutorResult;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ConditionResult = /** @class */ (function (_super) {
    __extends(ConditionResult, _super);
    function ConditionResult() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    ConditionResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.conditionExecutionResult = false;
        this.message = null;
    };
    /**
     * @return {?}
     */
    ConditionResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "conditionExecutionResult", this.conditionExecutionResult);
        BpmnModelJsonHelper.SetAttributeValue(obj, "message", this.message);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ConditionResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.conditionExecutionResult = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "conditionExecutionResult");
        this.message = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "message");
    };
    return ConditionResult;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var RpcVoChangeDetail = /** @class */ (function (_super) {
    __extends(RpcVoChangeDetail, _super);
    function RpcVoChangeDetail() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    RpcVoChangeDetail.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.voId = null;
        this.changeDetail = null;
    };
    /**
     * @return {?}
     */
    RpcVoChangeDetail.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "changeDetail", this.changeDetail);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RpcVoChangeDetail.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
        this.changeDetail = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "changeDetail");
    };
    return RpcVoChangeDetail;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * RPC映射结果。
 * @abstract
 */
var  /**
 * RPC映射结果。
 * @abstract
 */
RpcMappingResult = /** @class */ (function (_super) {
    __extends(RpcMappingResult, _super);
    function RpcMappingResult() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return RpcMappingResult;
}(RpcVoChangeDetail));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var RpcCreationMappingResult = /** @class */ (function (_super) {
    __extends(RpcCreationMappingResult, _super);
    function RpcCreationMappingResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.associations = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RpcCreationMappingResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.associations = new Array();
    };
    /**
     * @return {?}
     */
    RpcCreationMappingResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RpcCreationMappingResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
    };
    return RpcCreationMappingResult;
}(RpcMappingResult));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var RpcCreationRuleExecutorResult = /** @class */ (function (_super) {
    __extends(RpcCreationRuleExecutorResult, _super);
    function RpcCreationRuleExecutorResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.mappingResults = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RpcCreationRuleExecutorResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.creationRuleId = null;
        this.creationRuleInstanceId = null;
        this.bizFlowchartId = null;
        this.mappingResults = new Array();
    };
    /**
     * @return {?}
     */
    RpcCreationRuleExecutorResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingResults", this.mappingResults);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RpcCreationRuleExecutorResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.creationRuleInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.mappingResults = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingResults", RpcCreationMappingResult);
    };
    return RpcCreationRuleExecutorResult;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var FormBillCreationProcessExecutorResult = /** @class */ (function (_super) {
    __extends(FormBillCreationProcessExecutorResult, _super);
    function FormBillCreationProcessExecutorResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.pushCreationMode = PushCreationMode.CREATION;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormBillCreationProcessExecutorResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.conditionResult = null;
        this.creationRuleExecutorResult = null;
        this.pushCreationMode = PushCreationMode.CREATION;
        this.downBillSuCode = null;
    };
    /**
     * @return {?}
     */
    FormBillCreationProcessExecutorResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "conditionResult", this.conditionResult);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleExecutorResult", this.creationRuleExecutorResult);
        BpmnModelJsonHelper.SetAttributeValue(obj, "pushCreationMode", this.pushCreationMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillSuCode", this.downBillSuCode);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormBillCreationProcessExecutorResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.conditionResult = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "conditionResult", ConditionResult);
        this.creationRuleExecutorResult = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "creationRuleExecutorResult", RpcCreationRuleExecutorResult);
        /** @type {?} */
        var pushCreationMode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "pushCreationMode");
        if (pushCreationMode != null) {
            this.pushCreationMode = PushCreationMode[pushCreationMode];
        }
        this.downBillSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillSuCode");
    };
    return FormBillCreationProcessExecutorResult;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var RpcCreationRuleExecutorArgs = /** @class */ (function (_super) {
    __extends(RpcCreationRuleExecutorArgs, _super);
    function RpcCreationRuleExecutorArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.sourceVoEntities = new Array();
        _this.targetVoProxyEntities = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RpcCreationRuleExecutorArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.sourceVoEntities = new Array();
        this.creationRuleId = null;
        this.bizFlowchartId = null;
        this.sourceVoId = null;
        this.targetVoProxyId = null;
        this.targetVoProxyEntities = new Array();
    };
    /**
     * @return {?}
     */
    RpcCreationRuleExecutorArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntities", BpmnModelJsonHelper.StringListToJson(this.sourceVoEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyId", this.targetVoProxyId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoProxyEntities", BpmnModelJsonHelper.StringListToJson(this.targetVoProxyEntities));
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RpcCreationRuleExecutorArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.sourceVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.sourceVoEntities = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoEntities"));
        this.targetVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyId");
        this.targetVoProxyEntities = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoProxyEntities"));
    };
    return RpcCreationRuleExecutorArgs;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var RpcFireBizEventArgs = /** @class */ (function (_super) {
    __extends(RpcFireBizEventArgs, _super);
    function RpcFireBizEventArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.voChangeDetails = null;
        _this.voEntityDatas = null;
        _this.voEntityDataIds = new Array();
        _this.associations = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RpcFireBizEventArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.voId = null;
        this.voChangeDetails = null;
        this.voEntityDatas = null;
        this.voEntityDataIds = new Array();
        this.bizEventId = null;
        this.voFuncSessionId = null;
        this.bizFlowchartId = null;
        this.upperBeFuncSessionId = null;
        this.associations = new Array();
    };
    /**
     * @return {?}
     */
    RpcFireBizEventArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voId", this.voId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voChangeDetails", this.voChangeDetails);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voEntityDatas", this.voEntityDatas);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voEntityDataIds", BpmnModelJsonHelper.StringListToJsonArray(this.voEntityDataIds));
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizEventId", this.bizEventId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "voFuncSessionId", this.voFuncSessionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upperBeFuncSessionId", this.upperBeFuncSessionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RpcFireBizEventArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.voId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voId");
        this.voChangeDetails = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voChangeDetails");
        this.voEntityDatas = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voEntityDatas");
        this.voEntityDataIds = BpmnModelJsonHelper.StringListFromJsonArray(BpmnModelJsonHelper.GetAttributeValue(obj, "voEntityDataIds"));
        this.bizEventId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizEventId");
        this.voFuncSessionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "voFuncSessionId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.upperBeFuncSessionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upperBeFuncSessionId");
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", CreationRuleExecutorAssociations);
    };
    return RpcFireBizEventArgs;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var RpcWritebackMappingResult = /** @class */ (function (_super) {
    __extends(RpcWritebackMappingResult, _super);
    function RpcWritebackMappingResult() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return RpcWritebackMappingResult;
}(RpcMappingResult));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var RpcWritebackRuleExecutorArgs = /** @class */ (function (_super) {
    __extends(RpcWritebackRuleExecutorArgs, _super);
    function RpcWritebackRuleExecutorArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.sourceVoEntities = new Array();
        _this.associations = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RpcWritebackRuleExecutorArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.writebackRuleId = null;
        this.sourceVoId = null;
        this.targetVoFuncSessionId = null;
        this.sourceVoEntities = new Array();
        this.associations = new Array();
    };
    /**
     * @return {?}
     */
    RpcWritebackRuleExecutorArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "writebackRuleId", this.writebackRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoFuncSessionId", this.targetVoFuncSessionId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntities", BpmnModelJsonHelper.StringListToJson(this.sourceVoEntities));
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RpcWritebackRuleExecutorArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.writebackRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "writebackRuleId");
        this.sourceVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
        this.targetVoFuncSessionId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoFuncSessionId");
        this.sourceVoEntities = BpmnModelJsonHelper.StringListFromJson(BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoEntities"));
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
    };
    return RpcWritebackRuleExecutorArgs;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var RpcWritebackRuleExecutorResult = /** @class */ (function (_super) {
    __extends(RpcWritebackRuleExecutorResult, _super);
    function RpcWritebackRuleExecutorResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.mappingResults = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    RpcWritebackRuleExecutorResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.mappingResults = new Array();
    };
    /**
     * @return {?}
     */
    RpcWritebackRuleExecutorResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingResults", this.mappingResults);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RpcWritebackRuleExecutorResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.mappingResults = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingResults", RpcWritebackMappingResult);
    };
    return RpcWritebackRuleExecutorResult;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var AifManualCreationType = {
    Push: 'Push',
    Refer: 'Refer',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @deprecated please use \@edp-bif/runtime-api
 */
var  /**
 * @deprecated please use \@edp-bif/runtime-api
 */
AifRuntimeModuleConfig = /** @class */ (function () {
    function AifRuntimeModuleConfig() {
    }
    return AifRuntimeModuleConfig;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @deprecated please use \@edp-bif/runtime-api
 * @type {?}
 */
var SELECTOR_ARGS_INJECTOR = new InjectionToken("SELECTOR_ARGS_INJECTOR");

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
var  /**
 * @abstract
 */
FormRestArgs = /** @class */ (function (_super) {
    __extends(FormRestArgs, _super);
    function FormRestArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormRestArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.rpcCreationRuleExecutorResult = null;
    };
    /**
     * @return {?}
     */
    FormRestArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "rpcCreationRuleExecutorResult", this.rpcCreationRuleExecutorResult);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormRestArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.rpcCreationRuleExecutorResult = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "rpcCreationRuleExecutorResult");
    };
    return FormRestArgs;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var FormCreateArgs = /** @class */ (function (_super) {
    __extends(FormCreateArgs, _super);
    function FormCreateArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormCreateArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.formVoId = null;
    };
    /**
     * @return {?}
     */
    FormCreateArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "formVoId", this.formVoId);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormCreateArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.formVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formVoId");
    };
    return FormCreateArgs;
}(FormRestArgs));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
var  /**
 * @abstract
 */
FormRestResult = /** @class */ (function (_super) {
    __extends(FormRestResult, _super);
    function FormRestResult() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return FormRestResult;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var FormCreateResult = /** @class */ (function (_super) {
    __extends(FormCreateResult, _super);
    function FormCreateResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.entityDataList = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormCreateResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.entityDataList = new Array();
    };
    /**
     * @return {?}
     */
    FormCreateResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "entityDataList", BpmnModelJsonHelper.StringListToJsonArray(this.entityDataList));
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormCreateResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.entityDataList = BpmnModelJsonHelper.StringListFromJsonArray(BpmnModelJsonHelper.GetAttributeValue(obj, "entityDataList"));
    };
    return FormCreateResult;
}(FormRestResult));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 表单手工生单时执行生单规则参数类。
 */
var  /**
 * 表单手工生单时执行生单规则参数类。
 */
FormCreationRuleExecutorArgs = /** @class */ (function (_super) {
    __extends(FormCreationRuleExecutorArgs, _super);
    function FormCreationRuleExecutorArgs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * 来源单据ID树形列表
         */
        _this.sourceVoEntityDatas = new Array();
        /**
         * 目标单据主表ID列表
         */
        _this.targetVoDataIds = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormCreationRuleExecutorArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.sourceVoEntityDatas = new Array();
        this.creationRuleId = null;
        this.bizFlowchartId = null;
        this.targetVoDataIds = new Array();
    };
    /**
     * @return {?}
     */
    FormCreationRuleExecutorArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntityDatas", BpmnModelJsonHelper.StringListToJsonArray(this.sourceVoEntityDatas));
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoDataIds", BpmnModelJsonHelper.StringListToJsonArray(this.targetVoDataIds));
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormCreationRuleExecutorArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.sourceVoEntityDatas = BpmnModelJsonHelper.StringListFromJsonArray(BpmnModelJsonHelper.GetAttributeValue(obj, "sourceVoEntityDatas"));
        this.targetVoDataIds = BpmnModelJsonHelper.StringListFromJsonArray((/** @type {?} */ (BpmnModelJsonHelper.GetAttributeValue(obj, "targetVoDataIds"))));
    };
    return FormCreationRuleExecutorArgs;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var FormEntityData = /** @class */ (function (_super) {
    __extends(FormEntityData, _super);
    function FormEntityData() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.children = {};
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormEntityData.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.dataId = null;
        if (this.children != null) {
            for (var key in this.children) {
                delete this.children[key];
            }
        }
    };
    /**
     * @return {?}
     */
    FormEntityData.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "dataId", this.dataId);
        if (this.children != null) {
            /** @type {?} */
            var childrenObj = {};
            for (var key in this.children) {
                BpmnModelJsonHelper.SetAttributeValue(childrenObj, key, this.children[key]);
            }
            BpmnModelJsonHelper.SetAttributeValue(obj, "children", childrenObj);
        }
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormEntityData.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.dataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dataId");
        /** @type {?} */
        var childrenObj = BpmnModelJsonHelper.GetAttributeValue(obj, "children");
        if (childrenObj != null) {
            for (var key in childrenObj) {
                /** @type {?} */
                var value = BpmnModelJsonHelper.GetAttributeValueAsList(childrenObj, key, FormEntityData);
                this.children[key] = value;
            }
        }
    };
    return FormEntityData;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var FormUpdateArgs = /** @class */ (function (_super) {
    __extends(FormUpdateArgs, _super);
    function FormUpdateArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return FormUpdateArgs;
}(FormRestArgs));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var FormUpdateResult = /** @class */ (function (_super) {
    __extends(FormUpdateResult, _super);
    function FormUpdateResult() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return FormUpdateResult;
}(FormRestResult));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var FormGetCreationRuleEntityExArgs = /** @class */ (function (_super) {
    __extends(FormGetCreationRuleEntityExArgs, _super);
    function FormGetCreationRuleEntityExArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormGetCreationRuleEntityExArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.bizFlowchartId = null;
        this.sourceVoId = null;
        this.targetVoId = null;
    };
    /**
     * @return {?}
     */
    FormGetCreationRuleEntityExArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoId", this.sourceVoId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoId", this.targetVoId);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormGetCreationRuleEntityExArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.sourceVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "sourceVoId");
        this.targetVoId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "targetVoId");
    };
    return FormGetCreationRuleEntityExArgs;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var FormGetCreationRuleEntityExResult = /** @class */ (function (_super) {
    __extends(FormGetCreationRuleEntityExResult, _super);
    function FormGetCreationRuleEntityExResult() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.ruleEntityExes = new Array();
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormGetCreationRuleEntityExResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.ruleEntityExes = new Array();
    };
    /**
     * @return {?}
     */
    FormGetCreationRuleEntityExResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "ruleEntityExes", this.ruleEntityExes);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormGetCreationRuleEntityExResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.ruleEntityExes = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "ruleEntityExes", SlimCreationRuleEntityEx);
    };
    return FormGetCreationRuleEntityExResult;
}(JsonConvertableImpl));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var AifDataService = /** @class */ (function () {
    function AifDataService(repository) {
        this.getCreationRuleEntityByIdUrl = "/api/runtime/aif/v1.0/form/getcreationruleentityex";
        this.getCreationRuleEntityByIdMethod = "GET";
        this.getCreationRuleUrl = "/api/runtime/aif/v1.0/form/getcreationruleentityexes";
        this.getCreationRuleMethod = "POST";
        this.executeCreationRuleUrl = "/api/runtime/aif/v1.0/form/executecreationrule";
        this.executeCreationRuleMethod = "POST";
        this.formCreateUrl = "/api/runtime/aif/v1.0/form/create";
        this.formCreateMethod = "POST";
        this.formUpdateUrl = "/api/runtime/aif/v1.0/form/update";
        this.formUpdateMethod = "POST";
        this.generalTextUrl = "/api/runtime/aif/v1.0/generaltext";
        this.generalTextMethod = "GET";
        this.repository = (/** @type {?} */ (repository));
    }
    /**
     * 根据来源 VO Id获取生单规则实体列表。
     *
     * @param sourceVoId 来源 VO Id。
     * @return 符合条件的生单规则实体列表。
     */
    /**
     * 根据来源 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId 来源 VO Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    AifDataService.prototype.getSlimCreationRulesBySourceVoId = /**
     * 根据来源 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId 来源 VO Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    function (sourceVoId) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        return this.getCreationRuleList(args).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res.ruleEntityExes;
        })));
    };
    /**
     * 根据目标 VO Id获取生单规则实体列表。
     *
     * @param targetVoId 目标 VO Id。
     * @return 符合条件的生单规则实体列表。
     */
    /**
     * 根据目标 VO Id获取生单规则实体列表。
     *
     * @param {?} targetVoId 目标 VO Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    AifDataService.prototype.getSlimCreationRulesByTargetVoId = /**
     * 根据目标 VO Id获取生单规则实体列表。
     *
     * @param {?} targetVoId 目标 VO Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    function (targetVoId) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.targetVoId = targetVoId;
        return this.getCreationRuleList(args).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res.ruleEntityExes;
        })));
    };
    /**
     * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
     *
     * @param sourceVoId 来源 VO Id。
     * @param targetVoId 目标 VO Id。
     * @return 符合条件的生单规则实体列表。
     */
    /**
     * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId 来源 VO Id。
     * @param {?} targetVoId 目标 VO Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    AifDataService.prototype.getSlimCreationRulesBySourceVoIdAndTargetVoId = /**
     * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId 来源 VO Id。
     * @param {?} targetVoId 目标 VO Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    function (sourceVoId, targetVoId) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        args.targetVoId = targetVoId;
        return this.getCreationRuleList(args).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res.ruleEntityExes;
        })));
    };
    /**
     * 根据来源 VO Id获取生单规则实体列表。
     *
     * @param sourceVoId     来源 VO Id。
     * @param bizFlowchartId 业务场景图Id。
     * @return 符合条件的生单规则实体列表。
     */
    /**
     * 根据来源 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId     来源 VO Id。
     * @param {?} bizFlowchartId 业务场景图Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    AifDataService.prototype.getSlimCreationRulesBySourceVoIdAndFlowchartId = /**
     * 根据来源 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId     来源 VO Id。
     * @param {?} bizFlowchartId 业务场景图Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    function (sourceVoId, bizFlowchartId) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        args.bizFlowchartId = bizFlowchartId;
        return this.getCreationRuleList(args).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res.ruleEntityExes;
        })));
    };
    /**
     * 根据目标 VO Id获取生单规则实体列表。
     *
     * @param targetVoId     目标 VO Id。
     * @param bizFlowchartId 业务场景图Id。
     * @return 符合条件的生单规则实体列表。
     */
    /**
     * 根据目标 VO Id获取生单规则实体列表。
     *
     * @param {?} targetVoId     目标 VO Id。
     * @param {?} bizFlowchartId 业务场景图Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    AifDataService.prototype.getSlimCreationRulesByTargetVoIdAndFlowchartId = /**
     * 根据目标 VO Id获取生单规则实体列表。
     *
     * @param {?} targetVoId     目标 VO Id。
     * @param {?} bizFlowchartId 业务场景图Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    function (targetVoId, bizFlowchartId) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.targetVoId = targetVoId;
        args.bizFlowchartId = bizFlowchartId;
        return this.getCreationRuleList(args).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res.ruleEntityExes;
        })));
    };
    /**
     * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
     *
     * @param sourceVoId     来源 VO Id。
     * @param targetVoId     目标 VO Id。
     * @param bizFlowchartId 业务场景图Id。
     * @return 符合条件的生单规则实体列表。
     */
    /**
     * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId     来源 VO Id。
     * @param {?} targetVoId     目标 VO Id。
     * @param {?} bizFlowchartId 业务场景图Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    AifDataService.prototype.getSlimCreationRulesBySourceVoIdAndTargetVoIdAndFlowchartId = /**
     * 根据来源 VO Id和目标 VO Id获取生单规则实体列表。
     *
     * @param {?} sourceVoId     来源 VO Id。
     * @param {?} targetVoId     目标 VO Id。
     * @param {?} bizFlowchartId 业务场景图Id。
     * @return {?} 符合条件的生单规则实体列表。
     */
    function (sourceVoId, targetVoId, bizFlowchartId) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        args.targetVoId = targetVoId;
        args.bizFlowchartId = bizFlowchartId;
        return this.getCreationRuleList(args).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res.ruleEntityExes;
        })));
    };
    /**
     * 获取生单规则实体
     */
    /**
     * 获取生单规则实体
     * @param {?} creationRuleId
     * @return {?}
     */
    AifDataService.prototype.getSlimCreationRuleEntityExById = /**
     * 获取生单规则实体
     * @param {?} creationRuleId
     * @return {?}
     */
    function (creationRuleId) {
        /** @type {?} */
        var getCreationRuleById$ = this.repository.proxy.request(this.getCreationRuleEntityByIdUrl + "/" + creationRuleId, this.getCreationRuleEntityByIdMethod, null);
        return getCreationRuleById$;
    };
    /**
     * 获取表单配置实体
     * @param formSettingsId
     */
    /**
     * 获取表单配置实体
     * @param {?} formSettingsId
     * @return {?}
     */
    AifDataService.prototype.getFormSettingsById = /**
     * 获取表单配置实体
     * @param {?} formSettingsId
     * @return {?}
     */
    function (formSettingsId) {
        if (formSettingsId == null || formSettingsId.trim() == "") {
            throw "生单规则为空";
        }
        /** @type {?} */
        var formSettings$ = this.repository.proxy.request(this.generalTextUrl + "/" + formSettingsId, this.generalTextMethod, null, { responseType: "text" });
        return formSettings$.pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var generateText = new GeneralTextEntity();
            generateText.LoadFromJson(data);
            /** @type {?} */
            var formSettings = new FormSettings();
            formSettings.LoadFromJson(generateText.text);
            return formSettings;
        })));
    };
    /**
     * 执行生单规则
     */
    /**
     * 执行生单规则
     * @param {?} args
     * @return {?}
     */
    AifDataService.prototype.executeCreationRule = /**
     * 执行生单规则
     * @param {?} args
     * @return {?}
     */
    function (args) {
        /** @type {?} */
        var body = args.ConvertToJson();
        /** @type {?} */
        var executeCreationRule$ = this.repository.proxy.request(this.executeCreationRuleUrl, this.executeCreationRuleMethod, null, {
            body: body,
            responseType: 'text'
        });
        return executeCreationRule$.pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var res = new RpcCreationRuleExecutorResult();
            res.LoadFromJson(data);
            return res;
        })));
    };
    /**
     * 创建单据
     */
    /**
     * 创建单据
     * @param {?} args
     * @return {?}
     */
    AifDataService.prototype.createEntityDatas = /**
     * 创建单据
     * @param {?} args
     * @return {?}
     */
    function (args) {
        var _this = this;
        /** @type {?} */
        var body = args.ConvertToJson();
        /** @type {?} */
        var options = {
            body: body,
            responseType: 'text'
        };
        /** @type {?} */
        var create$ = this.repository.proxy.request(this.formCreateUrl, this.formCreateMethod, null, options);
        return create$.pipe(switchMap((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            /** @type {?} */
            var result = new FormCreateResult();
            result.LoadFromJson(res);
            /** @type {?} */
            var entities = result.entityDataList.map((/**
             * @param {?} entityStr
             * @return {?}
             */
            function (entityStr) {
                /** @type {?} */
                var newEntity = _this.repository.buildEntity(JSON.parse(entityStr));
                _this.appendInitialData(newEntity, newEntity);
                return newEntity;
            }));
            /** @type {?} */
            var origin = _this.repository.entityCollection.getAllEntities();
            if (origin == null || origin.length == 0) {
                _this.repository.entityCollection.loadEntities(entities);
            }
            else {
                _this.repository.entityCollection.addEntities(entities);
            }
            return of(result);
        })));
    };
    /**
     * 修改单据
     */
    /**
     * 修改单据
     * @param {?} args
     * @return {?}
     */
    AifDataService.prototype.updateEntityDatas = /**
     * 修改单据
     * @param {?} args
     * @return {?}
     */
    function (args) {
        /** @type {?} */
        var body = args.ConvertToJson();
        /** @type {?} */
        var options = {
            body: body,
            responseType: "text"
        };
        /** @type {?} */
        var update$ = this.repository.proxy.request(this.formUpdateUrl, this.formUpdateMethod, null, options);
        return update$.pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var res = new FormUpdateResult();
            res.LoadFromJson(data);
            return res;
        })));
    };
    /**
     * @private
     * @param {?} args
     * @return {?}
     */
    AifDataService.prototype.getCreationRuleList = /**
     * @private
     * @param {?} args
     * @return {?}
     */
    function (args) {
        if (args == null) {
            throw "args不可为空";
        }
        if (typeof args == "string") {
            if (args == "") {
                throw "args不可为空";
            }
            /** @type {?} */
            var temp = new FormGetCreationRuleEntityExArgs();
            temp.LoadFromJson(args);
            args = temp;
        }
        /** @type {?} */
        var getCreationRule$ = this.repository.proxy.request(this.getCreationRuleUrl, this.getCreationRuleMethod, null, { body: args.ConvertToJson(), responseType: "text" });
        return getCreationRule$.pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var result = new FormGetCreationRuleEntityExResult();
            result.LoadFromJson(data);
            return result;
        })));
    };
    /**
 * 参考farris/bef EntityUtil实现
 * @param entity
 * @param initialData
 */
    /**
     * 参考farris/bef EntityUtil实现
     * @private
     * @param {?} entity
     * @param {?} initialData
     * @return {?}
     */
    AifDataService.prototype.appendInitialData = /**
     * 参考farris/bef EntityUtil实现
     * @private
     * @param {?} entity
     * @param {?} initialData
     * @return {?}
     */
    function (entity, initialData) {
        /** @type {?} */
        var data = Object.assign({}, initialData);
        delete data.id;
        delete data.parentID;
        entity['initialData'] = data;
    };
    AifDataService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AifDataService.ctorParameters = function () { return [
        { type: Repository }
    ]; };
    return AifDataService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @deprecated please use \@edp-bif/runtime-api
 */
var ReferSelectorComponent = /** @class */ (function () {
    function ReferSelectorComponent(args, dataService, localeId) {
        var _this = this;
        this.dataService = dataService;
        //来源VoType列表
        this.sourceVoIdList = [];
        //来源VoType与来源VoName映射
        this.sourceVoNameDict = {};
        //来源VoType与相应的生单规则列表映射
        this.creationRuleDict = {};
        this.bsModalRef4AifSelector = null;
        this.subject4CreationRule = new Subject();
        this.clickOk4AifSelector = (/**
         * @return {?}
         */
        function () {
            _this.subject4CreationRule.next(_this.selectedCreationRule);
            _this.subject4CreationRule.complete();
            _this.bsModalRef4AifSelector.close();
        });
        this.clickClose4AifSelector = (/**
         * @return {?}
         */
        function () {
            _this.bsModalRef4AifSelector.close();
        });
        this.args = new FormGetCreationRuleEntityExArgs();
        this.args.LoadFromJsonObject(args);
    }
    Object.defineProperty(ReferSelectorComponent.prototype, "selectedSourceVoId", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selectedSourceVoId;
        },
        set: /**
         * @param {?} sourceVoId
         * @return {?}
         */
        function (sourceVoId) {
            this._selectedSourceVoId = sourceVoId;
            if (sourceVoId != null) {
                this.selectedCreationRule = this.creationRuleDict[this._selectedSourceVoId][0];
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ReferSelectorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.loadItems();
    };
    /**
     * @private
     * @return {?}
     */
    ReferSelectorComponent.prototype.loadItems = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var list$ = null;
        if (this.args.sourceVoId == null) {
            if (this.args.bizFlowchartId == null) {
                list$ = this.dataService.getSlimCreationRulesByTargetVoId(this.args.targetVoId);
            }
            else {
                list$ = this.dataService.getSlimCreationRulesByTargetVoIdAndFlowchartId(this.args.targetVoId, this.args.bizFlowchartId);
            }
        }
        else {
            if (this.args.bizFlowchartId == null) {
                list$ = this.dataService.getSlimCreationRulesBySourceVoIdAndTargetVoId(this.args.sourceVoId, this.args.targetVoId);
            }
            else {
                list$ = this.dataService.getSlimCreationRulesBySourceVoIdAndTargetVoIdAndFlowchartId(this.args.sourceVoId, this.args.targetVoId, this.args.bizFlowchartId);
            }
        }
        list$.subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            /** @type {?} */
            var list = result;
            list.forEach((/**
             * @param {?} elt
             * @return {?}
             */
            function (elt) {
                if (_this.sourceVoIdList.includes(elt.sourceVoId) === false) {
                    _this.sourceVoIdList.push(elt.sourceVoId);
                    _this.creationRuleDict[elt.sourceVoId] = new Array();
                }
                _this.sourceVoNameDict[elt.sourceVoId] = elt.sourceVoName;
                _this.creationRuleDict[elt.sourceVoId].push(elt);
            }));
            _this.selectedSourceVoId = _this.sourceVoIdList.length > 0 ? _this.sourceVoIdList[0] : null;
        }));
    };
    ReferSelectorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-refer-selector',
                    template: "<div class=\"g-main p-2\">\r\n  <!--  -->\r\n  <div class=\"form-group\" style=\"height: 150px;overflow-y: auto;\">\r\n    <div [hidden]=\"sourceVoIdList.length != 0\"><span style=\"color: red;\">{{ 'emptyResult' | aifLocale }}</span></div>\r\n    <div>{{ 'selectSourceBillType' | aifLocale }}</div>\r\n    <div *ngFor=\"let sourceVoId of sourceVoIdList;let i=index\" class=\"col-sm-12 col-md-5 form-check form-check-inline\">\r\n      <input class=\"form-check-input\" type=\"radio\" [name]=\"'sourceVoId'+i\" id=\"sourceVoId\" [value]=\"sourceVoId\"\r\n        [(ngModel)]=\"selectedSourceVoId\" />\r\n      <label class=\"form-check-label\" [for]=\"'sourceVoId'+i\">{{sourceVoNameDict[sourceVoId]}}</label>\r\n    </div>\r\n  </div>\r\n\r\n  <!--  -->\r\n  <div class=\"form-group\" style=\"\">\r\n    <div>{{ 'selectCreationRule' | aifLocale }}</div>\r\n    <select class=\"form-control\" name=\"selectedCreationRule\" [(ngModel)]=\"selectedCreationRule\">\r\n      <option *ngFor=\"let cr of creationRuleDict[selectedSourceVoId];\" [ngValue]=\"cr\">\r\n        {{cr.name}}</option>\r\n    </select>\r\n  </div>\r\n\r\n  <!--  -->\r\n  <div class=\"form-group\" style=\"\">\r\n    <div>{{ 'creationRuleDesciption' | aifLocale }}</div>\r\n    <textarea readonly class=\"form-control\" name=\"description\" rows=\"3\"\r\n      [ngModel]=\"selectedCreationRule&&selectedCreationRule.description\"></textarea>\r\n  </div>\r\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ReferSelectorComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [SELECTOR_ARGS_INJECTOR,] }] },
        { type: AifDataService },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    return ReferSelectorComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @deprecated please use \@edp-bif/runtime-api
 */
var PushSelectorComponent = /** @class */ (function () {
    function PushSelectorComponent(args, dataService, localeId) {
        var _this = this;
        this.dataService = dataService;
        //目标VoId列表
        this.targetVoIdList = [];
        //目标VoId与目标VoName映射
        this.targetVoNameDict = {};
        //目标VoId与相应的生单规则列表映射
        this.creationRuleDict = {};
        this.bsModalRef4AifSelector = null;
        this.subject4CreationRule = new Subject();
        this.clickOk4AifSelector = (/**
         * @return {?}
         */
        function () {
            _this.subject4CreationRule.next(_this.selectedCreationRule);
            _this.subject4CreationRule.complete();
            _this.bsModalRef4AifSelector.close();
        });
        this.clickClose4AifSelector = (/**
         * @return {?}
         */
        function () {
            _this.bsModalRef4AifSelector.close();
        });
        this.args = new FormGetCreationRuleEntityExArgs();
        this.args.LoadFromJsonObject(args);
    }
    Object.defineProperty(PushSelectorComponent.prototype, "selectedTargetVoId", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selectedTargetVoId;
        },
        set: /**
         * @param {?} targetVoId
         * @return {?}
         */
        function (targetVoId) {
            this._selectedTargetVoId = targetVoId;
            if (targetVoId != null) {
                this.selectedCreationRule = this.creationRuleDict[this._selectedTargetVoId][0];
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PushSelectorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.loadItems();
    };
    /**
     * @private
     * @return {?}
     */
    PushSelectorComponent.prototype.loadItems = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var list$ = null;
        if (this.args.targetVoId == null) {
            if (this.args.bizFlowchartId == null) {
                list$ = this.dataService.getSlimCreationRulesBySourceVoId(this.args.sourceVoId);
            }
            else {
                list$ = this.dataService.getSlimCreationRulesBySourceVoIdAndFlowchartId(this.args.sourceVoId, this.args.bizFlowchartId);
            }
        }
        else {
            if (this.args.bizFlowchartId == null) {
                list$ = this.dataService.getSlimCreationRulesBySourceVoIdAndTargetVoId(this.args.sourceVoId, this.args.targetVoId);
            }
            else {
                list$ = this.dataService.getSlimCreationRulesBySourceVoIdAndTargetVoIdAndFlowchartId(this.args.sourceVoId, this.args.targetVoId, this.args.bizFlowchartId);
            }
        }
        list$.subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            /** @type {?} */
            var list = result;
            list.forEach((/**
             * @param {?} elt
             * @return {?}
             */
            function (elt) {
                if (_this.targetVoIdList.includes(elt.targetVoId) === false) {
                    _this.targetVoIdList.push(elt.targetVoId);
                    _this.creationRuleDict[elt.targetVoId] = new Array();
                }
                _this.targetVoNameDict[elt.targetVoId] = elt.targetVoName;
                _this.creationRuleDict[elt.targetVoId].push(elt);
            }));
            _this.selectedTargetVoId = _this.targetVoIdList.length > 0 ? _this.targetVoIdList[0] : null;
        }));
    };
    PushSelectorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-push-selector',
                    template: "<div class=\"g-main p-2\">\r\n  <!--  -->\r\n  <div class=\"form-group\" style=\"height: 150px;overflow-y: auto;\">\r\n    <div [hidden]=\"targetVoIdList.length != 0\"><span style=\"color: red;\">{{ 'emptyResult' | aifLocale }}</span></div>\r\n    <div>{{ 'selectTargetBillType' | aifLocale }}</div>\r\n    <div *ngFor=\"let targetVoId of targetVoIdList;let i=index\" class=\"col-sm-12 col-md-5 form-check form-check-inline\">\r\n      <input class=\"form-check-input\" type=\"radio\" [name]=\"'targetVoId'+i\" id=\"targetVoId\" [value]=\"targetVoId\"\r\n        [(ngModel)]=\"selectedTargetVoId\" />\r\n      <label class=\"form-check-label\" [for]=\"'targetVoId'+i\">{{targetVoNameDict[targetVoId]}}</label>\r\n    </div>\r\n  </div>\r\n\r\n  <!--  -->\r\n  <div class=\"form-group\" style=\"\">\r\n    <div>{{ 'selectCreationRule' | aifLocale }}</div>\r\n    <select class=\"form-control\" name=\"selectedCreationRule\" [(ngModel)]=\"selectedCreationRule\">\r\n      <option *ngFor=\"let cr of creationRuleDict[selectedTargetVoId];\" [ngValue]=\"cr\">\r\n        {{cr.name}}</option>\r\n    </select>\r\n  </div>\r\n\r\n  <!--  -->\r\n  <div class=\"form-group\" style=\"\">\r\n    <div>{{ 'creationRuleDesciption' | aifLocale }}</div>\r\n    <textarea readonly class=\"form-control\" name=\"description\" rows=\"3\"\r\n      [ngModel]=\"selectedCreationRule&&selectedCreationRule.description\"></textarea>\r\n  </div>\r\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    PushSelectorComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [SELECTOR_ARGS_INJECTOR,] }] },
        { type: AifDataService },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    return PushSelectorComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @deprecated please use \@edp-bif/runtime-api
 */
var SourceDataSelectorComponent = /** @class */ (function () {
    function SourceDataSelectorComponent(localeId) {
        var _this = this;
        this.subject4EntityDatas = new Subject();
        this.cols = [];
        this.treedata = [];
        this.clickOk4AifSelector = (/**
         * @return {?}
         */
        function () {
            _this.subject4EntityDatas.next([]);
            _this.subject4EntityDatas.complete();
            _this.bsModalRef4AifSelector.close();
        });
        this.clickClose4AifSelector = (/**
         * @return {?}
         */
        function () {
            _this.bsModalRef4AifSelector.close();
        });
    }
    /**
     * @return {?}
     */
    SourceDataSelectorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    SourceDataSelectorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-source-data-selector',
                    template: "<div class=\"g-main p-2\">\r\n  <farris-treetable #treetable [columns]=\"cols\" [showBorder]=\"false\" [showHeader]=\"false\" [data]=\"treedata\"\r\n    [idField]=\"'id'\" [fit]=\"true\" [singleSelect]=\"true\" [showCheckbox]=\"false\" [showIcon]=\"true\" [disabled]=\"false\">\r\n  </farris-treetable>\r\n\r\n  <ng-template #cell let-ctx>\r\n\r\n  </ng-template>\r\n\r\n  <p>\r\n    Hello world\r\n  </p>\r\n\r\n  <!-- <div class=\"form-group\" style=\"float: right;\">\r\n    <button class=\"btn-primary btn btn-sm ml-2\" (click)=\"clickOk4Aif()\">\u786E\u5B9A</button>\r\n    <button class=\"btn btn-sm ml-2 btn-default\" (click)=\"clickClose4Aif()\">\u53D6\u6D88</button>\r\n  </div> -->\r\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    SourceDataSelectorComponent.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    SourceDataSelectorComponent.propDecorators = {
        treetable: [{ type: ViewChild, args: ["treetable",] }]
    };
    return SourceDataSelectorComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @deprecated please use \@edp-bif/runtime-api
 * @type {?}
 */
var LANG_RESOURCES = {
    "zh-CHS": {
        "confirm": "确定",
        "cancel": "取消",
        "emptyResult": "查询结果为空",
        "selectSourceBillType": "选择来源单据类型",
        "selectTargetBillType": "选择目标单据类型",
        "selectCreationRule": "选择生单规则",
        "creationRuleDesciption": "生单规则描述",
        "creationRuleSelector": "生单规则选择器",
        "sourceDataSelector": "来源单据选择器",
    },
    "en": {
        "confirm": "confirm",
        "cancel": "cancel",
        "emptyResult": "The result is empty",
        "selectSourceBillType": "Select source bill type",
        "selectTargetBillType": "Select target bill type",
        "selectCreationRule": "Select creation rule",
        "creationRuleDesciption": "Desciption",
        "creationRuleSelector": "CreationRule Selector",
        "sourceDataSelector": "SourceBill Selector",
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @deprecated please use \@edp-bif/runtime-api
 */
var AifLocalePipe = /** @class */ (function () {
    function AifLocalePipe(localeId) {
        this.langCode = localeId || "zh-CHS";
        this.lang = LANG_RESOURCES[this.langCode];
    }
    /**
     * @param {?} name
     * @return {?}
     */
    AifLocalePipe.prototype.transform = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        return this.lang[name] || name;
    };
    AifLocalePipe.decorators = [
        { type: Pipe, args: [{ name: 'aifLocale' },] }
    ];
    /** @nocollapse */
    AifLocalePipe.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    return AifLocalePipe;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ɵ0 = localStorage.getItem('languageCode') || 'zh-CHS', ɵ1 = {};
/**
 * @deprecated please use \@edp-bif/runtime-api
 */
var AifRuntimeModule = /** @class */ (function () {
    function AifRuntimeModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    AifRuntimeModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            "ngModule": AifRuntimeModule,
            "providers": [
                { provide: AifRuntimeModuleConfig, useValue: config || {} },
            ]
        };
    };
    AifRuntimeModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        HttpClientModule,
                        ModalModule.forRoot(),
                        NotifyModule.forRoot(),
                        MessagerModule.forRoot(),
                        FarrisDialogModule.forRoot(),
                        TreeTableModule,
                    ],
                    providers: [
                        { provide: LOCALE_ID, useValue: ɵ0 },
                        { provide: AifRuntimeModuleConfig, useValue: ɵ1 },
                        { provide: NgModuleFactoryLoader, useClass: SystemJsNgModuleLoader },
                    ],
                    declarations: [
                        ReferSelectorComponent,
                        PushSelectorComponent,
                        SourceDataSelectorComponent,
                        AifLocalePipe,
                    ],
                    exports: [
                        ReferSelectorComponent,
                        PushSelectorComponent,
                        SourceDataSelectorComponent,
                    ],
                    entryComponents: [
                        ReferSelectorComponent,
                        PushSelectorComponent,
                        SourceDataSelectorComponent,
                    ]
                },] }
    ];
    return AifRuntimeModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @deprecated please use \@edp-bif/runtime-api
 * @type {?}
 */
var AIF_UISTATE_KEY_OPEN_TYPE = "abdbaec8-0bbe-4894-acee-b8d493bc4e1c";
/**
 * @deprecated please use \@edp-bif/runtime-api
 * @type {?}
 */
var AIF_UISTATE_KEY_CREATION_RULE_RESULT = "5eb32009-1d86-43df-96c8-8d685aea4e66";

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
var AbstractAifUiService = /** @class */ (function () {
    function AbstractAifUiService(frameContext, modalService, injector, cfr, loader, localeId) {
        this.frameContext = frameContext;
        this.modalService = modalService;
        this.injector = injector;
        this.cfr = cfr;
        this.loader = loader;
        this.localeId = localeId;
        /**
         * 内置模态框配置
         */
        this.defaultModalOptions = {
            title: null,
            width: 650,
            height: 415,
            buttons: null,
            showButtons: null,
        };
        this.aifLocalePipe = new AifLocalePipe(localeId);
    }
    /**
     * @protected
     * @param {?} args
     * @param {?} type
     * @param {?=} modalOptions
     * @param {?=} customSelectorModuleUrl
     * @return {?}
     */
    AbstractAifUiService.prototype.openCreationRuleSelector = /**
     * @protected
     * @param {?} args
     * @param {?} type
     * @param {?=} modalOptions
     * @param {?=} customSelectorModuleUrl
     * @return {?}
     */
    function (args, type, modalOptions, customSelectorModuleUrl) {
        var _this = this;
        if (typeof modalOptions == "string") {
            if (modalOptions == "") {
                throw "modalOptions不可为空";
            }
            modalOptions = JSON.parse(modalOptions);
        }
        modalOptions = modalOptions == null ? this.defaultModalOptions : modalOptions;
        modalOptions.title = modalOptions.title == null ? this.aifLocalePipe.transform("creationRuleSelector") : modalOptions.title;
        /** @type {?} */
        var cmpR = null;
        if (customSelectorModuleUrl == null || customSelectorModuleUrl.trim() == "") {
            /** @type {?} */
            var cmpF = this.cfr.resolveComponentFactory(type);
            /** @type {?} */
            var inj = ReflectiveInjector.resolveAndCreate([
                { provide: SELECTOR_ARGS_INJECTOR, useValue: args.ConvertToJsonObject() },
            ], this.injector);
            cmpR = cmpF.create(inj);
            this.configModalButtons(modalOptions, cmpR);
            cmpR.instance.bsModalRef4AifSelector = this.modalService.show(cmpR, modalOptions);
            return cmpR.instance.subject4CreationRule.asObservable();
        }
        else {
            return from(this.loader.load(customSelectorModuleUrl)).pipe(switchMap((/**
             * @param {?} moduleFactory
             * @return {?}
             */
            function (moduleFactory) {
                /** @type {?} */
                var moduleRef = moduleFactory.create(_this.frameContext.root.injector);
                cmpR = moduleRef.instance.createCreationRuleSelector(_this.frameContext.root.injector);
                _this.configModalButtons(modalOptions, cmpR);
                cmpR.instance.bsModalRef4AifSelector = _this.modalService.show(cmpR, modalOptions);
                return cmpR.instance.subject4CreationRule.asObservable();
            })));
        }
    };
    /**
     * 配置模态框按钮事件
     * @param modalOptions
     * @param cmpR
     */
    /**
     * 配置模态框按钮事件
     * @protected
     * @param {?} modalOptions
     * @param {?} cmpR
     * @return {?}
     */
    AbstractAifUiService.prototype.configModalButtons = /**
     * 配置模态框按钮事件
     * @protected
     * @param {?} modalOptions
     * @param {?} cmpR
     * @return {?}
     */
    function (modalOptions, cmpR) {
        modalOptions.buttons = [
            {
                text: this.aifLocalePipe.transform("confirm"),
                cls: "btn btn-primary",
                handle: cmpR.instance.clickOk4AifSelector,
            },
            {
                text: this.aifLocalePipe.transform("cancel"),
                cls: "btn btn-secondary",
                handle: cmpR.instance.clickClose4AifSelector,
            }
        ];
        modalOptions.showButtons = true;
    };
    /** @nocollapse */
    AbstractAifUiService.ctorParameters = function () { return [
        { type: FrameContext },
        { type: BsModalService },
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: NgModuleFactoryLoader },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    return AbstractAifUiService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var AifSourceUiService = /** @class */ (function (_super) {
    __extends(AifSourceUiService, _super);
    function AifSourceUiService(aifDataService, navigationService, frameContext, modalService, injector, cfr, loader, localeId) {
        var _this = _super.call(this, frameContext, modalService, injector, cfr, loader, localeId) || this;
        _this.aifDataService = aifDataService;
        _this.navigationService = navigationService;
        _this.frameContext = frameContext;
        _this.modalService = modalService;
        _this.injector = injector;
        _this.cfr = cfr;
        _this.loader = loader;
        _this.localeId = localeId;
        _this.aifUIStateCreationRuleResultKey = AIF_UISTATE_KEY_CREATION_RULE_RESULT;
        return _this;
    }
    /**
     * 下推时打开生单规则选择器
     */
    /**
     * 下推时打开生单规则选择器
     * @param {?} sourceVoId
     * @param {?=} bizFlowchartId
     * @param {?=} modalOptions
     * @param {?=} customSelectorModuleUrl
     * @return {?}
     */
    AifSourceUiService.prototype.openCreationRuleSelector4Push = /**
     * 下推时打开生单规则选择器
     * @param {?} sourceVoId
     * @param {?=} bizFlowchartId
     * @param {?=} modalOptions
     * @param {?=} customSelectorModuleUrl
     * @return {?}
     */
    function (sourceVoId, bizFlowchartId, modalOptions, customSelectorModuleUrl) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        args.bizFlowchartId = bizFlowchartId;
        return this.openCreationRuleSelector(args, PushSelectorComponent, modalOptions, customSelectorModuleUrl);
    };
    /**
     * 特定下推时打开生单规则选择器
     */
    /**
     * 特定下推时打开生单规则选择器
     * @param {?} sourceVoId
     * @param {?} targetVoId
     * @param {?=} bizFlowchartId
     * @param {?=} modalOptions
     * @param {?=} customSelectorModuleUrl
     * @return {?}
     */
    AifSourceUiService.prototype.openCreationRuleSelector4SpecificPush = /**
     * 特定下推时打开生单规则选择器
     * @param {?} sourceVoId
     * @param {?} targetVoId
     * @param {?=} bizFlowchartId
     * @param {?=} modalOptions
     * @param {?=} customSelectorModuleUrl
     * @return {?}
     */
    function (sourceVoId, targetVoId, bizFlowchartId, modalOptions, customSelectorModuleUrl) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        args.targetVoId = targetVoId;
        args.bizFlowchartId = bizFlowchartId;
        return this.openCreationRuleSelector(args, PushSelectorComponent, modalOptions, customSelectorModuleUrl);
    };
    /**
     * 打开下游单据制单界面
     * @param formSettings
     * @param rpcCreationRuleExecutorResult
     * @param extParams
     */
    /**
     * 打开下游单据制单界面
     * @param {?} formSettings
     * @param {?} rpcCreationRuleExecutorResult
     * @param {?=} extParams
     * @return {?}
     */
    AifSourceUiService.prototype.openTargetForm = /**
     * 打开下游单据制单界面
     * @param {?} formSettings
     * @param {?} rpcCreationRuleExecutorResult
     * @param {?=} extParams
     * @return {?}
     */
    function (formSettings, rpcCreationRuleExecutorResult, extParams) {
        var e_1, _a;
        if (formSettings == null) {
            throw "参数formSettings不可为空";
        }
        if (rpcCreationRuleExecutorResult == null) {
            throw "参数rpcCreationRuleExecutorResult不可为空";
        }
        if (typeof rpcCreationRuleExecutorResult == "string") {
            if (rpcCreationRuleExecutorResult == "") {
                throw "参数rpcCreationRuleExecutorResult不可为空";
            }
            /** @type {?} */
            var temp = new RpcCreationRuleExecutorResult();
            temp.LoadFromJson(rpcCreationRuleExecutorResult);
            rpcCreationRuleExecutorResult = temp;
        }
        if ((rpcCreationRuleExecutorResult instanceof RpcCreationRuleExecutorResult) == false) {
            /** @type {?} */
            var res = new RpcCreationRuleExecutorResult();
            res.LoadFromJsonObject(rpcCreationRuleExecutorResult);
            rpcCreationRuleExecutorResult = res;
        }
        if (rpcCreationRuleExecutorResult.mappingResults == null || rpcCreationRuleExecutorResult.mappingResults.length == 0) {
            return;
        }
        else if (rpcCreationRuleExecutorResult.mappingResults.length == 1) {
            //只有一条数据则打开卡片
            this.navigationService.openApp(rpcCreationRuleExecutorResult.mappingResults[0].associations[0].tarDataId, formSettings.targetCardAppId, formSettings.targetCardAppEntrance, this.buildRouteParams(rpcCreationRuleExecutorResult, extParams == null ? {} : extParams), false);
        }
        else {
            if (formSettings.targetListAppId != null) {
                // 多条数据且配置批量编辑界面
                throw "暂不支持打开批量编辑界面";
                this.navigationService.openApp(rpcCreationRuleExecutorResult.mappingResults[0].associations[0].tarDataId, formSettings.targetListAppId, formSettings.targetListAppEntrance, this.buildRouteParams(rpcCreationRuleExecutorResult, extParams == null ? {} : extParams), false);
            }
            else {
                try {
                    //有多条数据但未配置批量编辑界面则打开多个卡片
                    for (var _b = __values(rpcCreationRuleExecutorResult.mappingResults), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var mappingResult = _c.value;
                        /** @type {?} */
                        var singleRes = new RpcCreationRuleExecutorResult();
                        singleRes.bizFlowchartId = rpcCreationRuleExecutorResult.bizFlowchartId;
                        singleRes.creationRuleId = rpcCreationRuleExecutorResult.creationRuleId;
                        singleRes.creationRuleInstanceId = rpcCreationRuleExecutorResult.creationRuleInstanceId;
                        singleRes.mappingResults = [mappingResult];
                        this.navigationService.openApp(mappingResult.associations[0].tarDataId, formSettings.targetCardAppId, formSettings.targetCardAppEntrance, this.buildRouteParams(singleRes, extParams == null ? {} : extParams), false);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
        }
    };
    /**
     *
     * @param action 目标App动作
     * @param pairs 其他实体参数键值对
     */
    /**
     *
     * @private
     * @param {?} rpcCreationRuleExecutorResult
     * @param {?} extParams
     * @return {?}
     */
    AifSourceUiService.prototype.buildRouteParams = /**
     *
     * @private
     * @param {?} rpcCreationRuleExecutorResult
     * @param {?} extParams
     * @return {?}
     */
    function (rpcCreationRuleExecutorResult, extParams) {
        /** @type {?} */
        var params = extParams;
        params[this.aifUIStateCreationRuleResultKey] = rpcCreationRuleExecutorResult;
        return params;
    };
    AifSourceUiService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AifSourceUiService.ctorParameters = function () { return [
        { type: AifDataService },
        { type: NavigationService },
        { type: FrameContext },
        { type: BsModalService },
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: NgModuleFactoryLoader },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    return AifSourceUiService;
}(AbstractAifUiService));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var AifTargetUiService = /** @class */ (function (_super) {
    __extends(AifTargetUiService, _super);
    function AifTargetUiService(aifDataService, frameContext, modalService, injector, cfr, loader, localeId) {
        var _this = _super.call(this, frameContext, modalService, injector, cfr, loader, localeId) || this;
        _this.aifDataService = aifDataService;
        _this.frameContext = frameContext;
        _this.modalService = modalService;
        _this.injector = injector;
        _this.cfr = cfr;
        _this.loader = loader;
        _this.localeId = localeId;
        _this.aifUIStateCreationRuleResultKey = AIF_UISTATE_KEY_CREATION_RULE_RESULT;
        return _this;
    }
    /**
     * 参照时打开生单规则选择器
     */
    /**
     * 参照时打开生单规则选择器
     * @param {?} targetVoId
     * @param {?=} bizFlowchartId
     * @param {?=} modalOptions
     * @param {?=} customSelectorModuleUrl
     * @return {?}
     */
    AifTargetUiService.prototype.openCreationRuleSelector4Refer = /**
     * 参照时打开生单规则选择器
     * @param {?} targetVoId
     * @param {?=} bizFlowchartId
     * @param {?=} modalOptions
     * @param {?=} customSelectorModuleUrl
     * @return {?}
     */
    function (targetVoId, bizFlowchartId, modalOptions, customSelectorModuleUrl) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.targetVoId = targetVoId;
        args.bizFlowchartId = bizFlowchartId;
        return this.openCreationRuleSelector(args, ReferSelectorComponent, modalOptions, customSelectorModuleUrl);
    };
    /**
     * 特定参照时打开生单规则选择器
     */
    /**
     * 特定参照时打开生单规则选择器
     * @param {?} sourceVoId
     * @param {?} targetVoId
     * @param {?=} bizFlowchartId
     * @param {?=} modalOptions
     * @param {?=} customSelectorModuleUrl
     * @return {?}
     */
    AifTargetUiService.prototype.openCreationRuleSelector4SpecificRefer = /**
     * 特定参照时打开生单规则选择器
     * @param {?} sourceVoId
     * @param {?} targetVoId
     * @param {?=} bizFlowchartId
     * @param {?=} modalOptions
     * @param {?=} customSelectorModuleUrl
     * @return {?}
     */
    function (sourceVoId, targetVoId, bizFlowchartId, modalOptions, customSelectorModuleUrl) {
        /** @type {?} */
        var args = new FormGetCreationRuleEntityExArgs();
        args.sourceVoId = sourceVoId;
        args.targetVoId = targetVoId;
        args.bizFlowchartId = bizFlowchartId;
        return this.openCreationRuleSelector(args, ReferSelectorComponent, modalOptions, customSelectorModuleUrl);
    };
    /**
     * 打开来源单据选择器
     * @param formSettings
     * @param modalOptions
     */
    /**
     * 打开来源单据选择器
     * @param {?=} formSettings
     * @param {?=} modalOptions
     * @return {?}
     */
    AifTargetUiService.prototype.openSourceDataSelector4Refer = /**
     * 打开来源单据选择器
     * @param {?=} formSettings
     * @param {?=} modalOptions
     * @return {?}
     */
    function (formSettings, modalOptions) {
        var _this = this;
        modalOptions = modalOptions == null ? this.defaultModalOptions : modalOptions;
        modalOptions.title = modalOptions.title == null ? this.aifLocalePipe.transform("sourceDataSelector") : modalOptions.title;
        /** @type {?} */
        var cmpR = null;
        if (formSettings == null || formSettings.sourceDataSelectorModuleUrl == null) {
            throw "暂无默认实现，请使用自定义来源单据选择器";
            /** @type {?} */
            var cmpF = this.cfr.resolveComponentFactory(SourceDataSelectorComponent);
            /** @type {?} */
            var inj = Injector.create([], this.injector);
            cmpR = cmpF.create(inj);
            this.configModalButtons(modalOptions, cmpR);
            cmpR.instance.bsModalRef4AifSelector = this.modalService.show(cmpR, modalOptions);
            return cmpR.instance.subject4EntityDatas.asObservable();
        }
        else {
            // return from(System.import(formSettings.sourceDataSelectorModuleUrl.substring(0, formSettings.sourceDataSelectorModuleUrl.indexOf("#")) + ".js")).pipe(
            return from(this.loader.load(formSettings.sourceDataSelectorModuleUrl)).pipe(switchMap((/**
             * @param {?} moduleFactory
             * @return {?}
             */
            function (moduleFactory) {
                /** @type {?} */
                var moduleRef = moduleFactory.create(_this.frameContext.root.injector);
                cmpR = moduleRef.instance.createSourceDataSelector(_this.frameContext.root.injector);
                _this.configModalButtons(modalOptions, cmpR);
                cmpR.instance.bsModalRef4AifSelector = _this.modalService.show(cmpR, modalOptions);
                return cmpR.instance.subject4EntityDatas.asObservable();
            })));
        }
    };
    /**
     * 获取传到根组件UIState上的生单规则执行器结果
     */
    /**
     * 获取传到根组件UIState上的生单规则执行器结果
     * @return {?}
     */
    AifTargetUiService.prototype.getRpcCreationRuleExecutorResult = /**
     * 获取传到根组件UIState上的生单规则执行器结果
     * @return {?}
     */
    function () {
        /** @type {?} */
        var temp = this.frameContext.root.uiState[this.aifUIStateCreationRuleResultKey];
        /** @type {?} */
        var result = new RpcCreationRuleExecutorResult();
        if (typeof temp == "string") {
            result.LoadFromJson(temp);
        }
        else if (typeof temp == "object") {
            result.LoadFromJsonObject(temp);
        }
        else {
            throw "不合法的类型";
        }
        return result;
    };
    AifTargetUiService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AifTargetUiService.ctorParameters = function () { return [
        { type: AifDataService },
        { type: FrameContext },
        { type: BsModalService },
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: NgModuleFactoryLoader },
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    return AifTargetUiService;
}(AbstractAifUiService));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { AifManualCreationType, AifRuntimeModuleConfig, AifRuntimeModule, AifDataService, AifSourceUiService, AifTargetUiService, SourceDataSelectorComponent, PushSelectorComponent, ReferSelectorComponent, BefContentVariable, CreationMappingResult, CreationRuleExecutorArgs, CreationRuleExecutorAssociations, CreationRuleExecutorResult, MappingAssociation, MappingResult, ConditionResult, FormBillCreationProcessExecutorResult, RpcCreationMappingResult, RpcCreationRuleExecutorArgs, RpcCreationRuleExecutorResult, RpcFireBizEventArgs, RpcMappingResult, RpcVoChangeDetail, RpcWritebackMappingResult, RpcWritebackRuleExecutorArgs, RpcWritebackRuleExecutorResult, FormCreateArgs, FormCreateResult, FormCreationRuleExecutorArgs, FormEntityData, FormRestArgs, FormRestResult, FormUpdateArgs, FormUpdateResult, FormGetCreationRuleEntityExArgs, FormGetCreationRuleEntityExResult, AIF_UISTATE_KEY_OPEN_TYPE, AIF_UISTATE_KEY_CREATION_RULE_RESULT, SELECTOR_ARGS_INJECTOR as ɵa, AifLocalePipe as ɵb, AbstractAifUiService as ɵc };
