// @ecp-caf/caf-common/esm2015/lib/cache/services/storage/cache.storage.abstract.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Abstract cache storage
 * @abstract
 */
export class CacheStorageAbstract {
}
if (false) {
    /**
     * Get item from storage
     * @abstract
     * @param {?} key
     * @return {?}
     */
    CacheStorageAbstract.prototype.getItem = function (key) { };
    /**
     * Set item to storage
     * @abstract
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    CacheStorageAbstract.prototype.setItem = function (key, value) { };
    /**
     * Remove item from storage
     * @abstract
     * @param {?} key
     * @return {?}
     */
    CacheStorageAbstract.prototype.removeItem = function (key) { };
    /**
     * Clear item in storage
     * @abstract
     * @return {?}
     */
    CacheStorageAbstract.prototype.clear = function () { };
    /**
     * Get current storage type
     * @abstract
     * @return {?}
     */
    CacheStorageAbstract.prototype.type = function () { };
    /**
     * Check if storage is enabled
     * @abstract
     * @return {?}
     */
    CacheStorageAbstract.prototype.isEnabled = function () { };
}
