// @gsp-wf/wf-process-editor/esm2015/lib/views/aif-process-list-help/services/process-list-help.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { map } from 'rxjs/operators';
import { FrameworkService, AppService } from '@gsp-sys/rtf-common';
import { BifMonitorConstant } from '../entity/BifMonitorConstant';
import { RtBillProcessListResult } from '../entity/RtBillProcessListResult';
import { RtShowEntity } from '../entity/RtShowEntity';
export class ProcessListHelpService {
    /**
     * @param {?} appService
     * @param {?} frameworkService
     * @param {?} httpService
     */
    constructor(appService, frameworkService, httpService) {
        this.appService = appService;
        this.frameworkService = frameworkService;
        this.httpService = httpService;
        this.baseUrl = "/api/runtime/aif/v1.0/rtbillprocess";
    }
    /**
     * @param {?} entity
     * @return {?}
     */
    openBifManagerApp(entity) {
        /** @type {?} */
        const map = new Map();
        map.set("id", entity.id);
        map.set("type", entity.bifMetadataType);
        map.set("srcFuncId", BifMonitorConstant.bifDiagnostorFuncId);
        /** @type {?} */
        const opt = {
            funcId: BifMonitorConstant.bifManagerFuncId,
            appId: BifMonitorConstant.bifMonitorAppId,
            appEntrance: BifMonitorConstant.bifManagerAppEntrance,
            appType: "menu",
            entityParams: {},
            queryStringParams: map,
            tabId: entity.id,
            tabName: entity.name
        };
        this.frameworkService.openMenu(opt);
    }
    /**
     * @param {?} args
     * @return {?}
     */
    getDatagridData(args) {
        /** @type {?} */
        const body = args.ConvertToJson();
        /** @type {?} */
        let data$ = null;
        //有业务种类参数时不分页
        if (args.bizType != null) {
            data$ = this.httpService.post(`${this.baseUrl}/getpatch`, body);
            return data$.pipe(map((/**
             * @param {?} res
             * @return {?}
             */
            (res) => {
                /** @type {?} */
                const listRes = new RtBillProcessListResult();
                listRes.entities = res.map((/**
                 * @param {?} elt
                 * @return {?}
                 */
                elt => {
                    /** @type {?} */
                    const e = new RtShowEntity();
                    e.LoadFromJsonObject(elt);
                    return e;
                }));
                return listRes;
            })));
        }
        else {
            data$ = this.httpService.post(`${this.baseUrl}/getpagedpatch`, body);
            return data$.pipe(map((/**
             * @param {?} res
             * @return {?}
             */
            (res) => {
                /** @type {?} */
                const listRes = new RtBillProcessListResult();
                listRes.LoadFromJsonObject(res);
                return listRes;
            })));
        }
    }
    /**
     * 根据流程编号查询流程名称
     * @param {?} code
     * @return {?}
     */
    getProcessNameByCode(code) {
        return this.httpService.request("GET", `/api/runtime/aif/v1.0/rtbillcreationprocess/getname/${code}`, { responseType: 'text' });
    }
}
ProcessListHelpService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ProcessListHelpService.ctorParameters = () => [
    { type: AppService },
    { type: FrameworkService },
    { type: HttpService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ProcessListHelpService.prototype.baseUrl;
    /**
     * @type {?}
     * @private
     */
    ProcessListHelpService.prototype.appService;
    /**
     * @type {?}
     * @private
     */
    ProcessListHelpService.prototype.frameworkService;
    /**
     * @type {?}
     * @private
     */
    ProcessListHelpService.prototype.httpService;
}
