// @edp-aif/runtime-api/esm2015/lib/entity/CreationMappingResult.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { MappingResult } from './MappingResult';
import { MappingAssociation } from './MappingAssociation';
import { BpmnModelJsonHelper } from '@edp-aif/common-api';
export class CreationMappingResult extends MappingResult {
    constructor() {
        super(...arguments);
        this.associations = new Array();
        this.entityData = null;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.associations = new Array();
        this.entityData = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "associations", this.associations);
        BpmnModelJsonHelper.SetAttributeValue(obj, "entityData", this.entityData);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.associations = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "associations", MappingAssociation);
        this.entityData = BpmnModelJsonHelper.GetAttributeValue(obj, "entityData");
    }
}
if (false) {
    /** @type {?} */
    CreationMappingResult.prototype.associations;
    /** @type {?} */
    CreationMappingResult.prototype.entityData;
}
