// @edp-aif/runtime-api/esm2015/lib/entity/CreationRuleExecutorResult.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
import { CreationMappingResult } from './CreationMappingResult';
export class CreationRuleExecutorResult extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.isAllAssociation = false;
        this.mappingResults = new Array();
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.creationRuleId = null;
        this.creationRuleInstanceId = null;
        this.bizFlowchartId = null;
        this.isAllAssociation = false;
        this.mappingResults = new Array();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isAllAssociation", this.isAllAssociation);
        BpmnModelJsonHelper.SetAttributeValue(obj, "mappingResults", this.mappingResults);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.creationRuleInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.isAllAssociation = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isAllAssociation");
        this.mappingResults = BpmnModelJsonHelper.GetAttributeValueAsList(obj, "mappingResults", CreationMappingResult);
    }
}
if (false) {
    /** @type {?} */
    CreationRuleExecutorResult.prototype.creationRuleId;
    /** @type {?} */
    CreationRuleExecutorResult.prototype.creationRuleInstanceId;
    /** @type {?} */
    CreationRuleExecutorResult.prototype.bizFlowchartId;
    /** @type {?} */
    CreationRuleExecutorResult.prototype.isAllAssociation;
    /** @type {?} */
    CreationRuleExecutorResult.prototype.mappingResults;
}
