// @edp-aif/runtime-api/esm2015/lib/entity/FormBillCreationProcessExecutorResult.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper, PushCreationMode } from "@edp-aif/common-api";
import { ConditionResult } from './ConditionResult';
import { RpcCreationRuleExecutorResult } from "./rpc/RpcCreationRuleExecutorResult";
export class FormBillCreationProcessExecutorResult extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.pushCreationMode = PushCreationMode.CREATION;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.conditionResult = null;
        this.creationRuleExecutorResult = null;
        this.pushCreationMode = PushCreationMode.CREATION;
        this.downBillSuCode = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "conditionResult", this.conditionResult);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleExecutorResult", this.creationRuleExecutorResult);
        BpmnModelJsonHelper.SetAttributeValue(obj, "pushCreationMode", this.pushCreationMode);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillSuCode", this.downBillSuCode);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.conditionResult = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "conditionResult", ConditionResult);
        this.creationRuleExecutorResult = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "creationRuleExecutorResult", RpcCreationRuleExecutorResult);
        /** @type {?} */
        let pushCreationMode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "pushCreationMode");
        if (pushCreationMode != null) {
            this.pushCreationMode = PushCreationMode[pushCreationMode];
        }
        this.downBillSuCode = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillSuCode");
    }
}
if (false) {
    /** @type {?} */
    FormBillCreationProcessExecutorResult.prototype.conditionResult;
    /** @type {?} */
    FormBillCreationProcessExecutorResult.prototype.creationRuleExecutorResult;
    /** @type {?} */
    FormBillCreationProcessExecutorResult.prototype.pushCreationMode;
    /** @type {?} */
    FormBillCreationProcessExecutorResult.prototype.downBillSuCode;
}
