// @gsp-wf/wf-dimension-def/esm2015/lib/refModule/panel/panel/panel-item-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation, Output, EventEmitter, ElementRef, Renderer2, ViewChild } from '@angular/core';
export class WFDimensionPanelComponent {
    /**
     * @param {?} render
     */
    constructor(render) {
        this.render = render;
        /* 自定义className */
        this.className = [];
        /* 面板是否可收折  默认收折*/
        this.foldable = true;
        // 设置手风琴模式  至多只有一个panel可以被选中
        this.accordion = false;
        this.modelChange = new EventEmitter();
        // @HostBinding('class.farris-panel') isPanel =
        // 双向绑定  当model数组改变时，相应的active状态的panel也要改变 从而改变视图
        // 双向绑定 用于存取active状态的panel
        this.modelValue = [];
        this.subscriber = [];
    }
    /* panel数据存储 */
    /**
     * @return {?}
     */
    get model() {
        return this.modelValue;
    }
    /**
     * @param {?} modelValue
     * @return {?}
     */
    set model(modelValue) {
        // 获取用户输入的值
        this.modelValue = modelValue;
        this.updateSubscriber();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        // 添加class名
        if (this.className.length) {
            this.className.forEach((/**
             * @param {?} ele
             * @return {?}
             */
            ele => {
                this.render.addClass(this.farrisPanel.nativeElement, ele);
            }));
        }
        this.updateSubscriber();
    }
    /**
     * @param {?} value panel存储激活值
     * @return {?}
     */
    updateModel(value) {
        /** @type {?} */
        const index = this.modelValue.findIndex((/**
         * @param {?} val
         * @return {?}
         */
        val => val === value));
        // 如果没有匹配到点击的panel值  可能model里面不存在 可能model为空
        if (index < 0) {
            // 若是手风琴模式  先把所有panel处于未激活状态
            if (this.accordion) {
                this.model = [];
            }
            // 如果不是手风琴模式  直接激活即可
            this.modelValue.push(value);
            this.updateSubscriber();
            return this.modelChange.emit(this.modelValue);
        }
        // 如果index不为-1
        // 手风琴模式  关闭所有激活状态
        if (this.accordion) {
            this.model = [];
        }
        else {
            // 非手风琴模式 直接关闭当前panel即可
            this.modelValue.splice(index, 1);
        }
        this.updateSubscriber();
        this.modelChange.emit(this.modelValue);
    }
    /* 函数分发 */
    /**
     * @private
     * @return {?}
     */
    updateSubscriber() {
        this.subscriber.forEach((/**
         * @param {?} handle
         * @return {?}
         */
        handle => handle()));
    }
}
WFDimensionPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'wf-dimension-panel',
                template: "<div #farrisPanel class=\"farris-panel\" [class.accordion]=\"accordion\" [style.width]=\"width+'px'\" [style.height]=\"height+'px'\">\r\n    <ng-content>\r\n    </ng-content>\r\n</div>",
                encapsulation: ViewEncapsulation.None,
                styles: [".farris-panel{margin:25px}"]
            }] }
];
/** @nocollapse */
WFDimensionPanelComponent.ctorParameters = () => [
    { type: Renderer2 }
];
WFDimensionPanelComponent.propDecorators = {
    farrisPanel: [{ type: ViewChild, args: ['farrisPanel',] }],
    width: [{ type: Input }],
    height: [{ type: Input }],
    className: [{ type: Input }],
    foldable: [{ type: Input }],
    defaultExpand: [{ type: Input }],
    accordion: [{ type: Input }],
    model: [{ type: Input }],
    modelChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    WFDimensionPanelComponent.prototype.farrisPanel;
    /** @type {?} */
    WFDimensionPanelComponent.prototype.width;
    /** @type {?} */
    WFDimensionPanelComponent.prototype.height;
    /** @type {?} */
    WFDimensionPanelComponent.prototype.className;
    /** @type {?} */
    WFDimensionPanelComponent.prototype.foldable;
    /** @type {?} */
    WFDimensionPanelComponent.prototype.defaultExpand;
    /** @type {?} */
    WFDimensionPanelComponent.prototype.accordion;
    /** @type {?} */
    WFDimensionPanelComponent.prototype.modelChange;
    /** @type {?} */
    WFDimensionPanelComponent.prototype.modelValue;
    /** @type {?} */
    WFDimensionPanelComponent.prototype.subscriber;
    /**
     * @type {?}
     * @private
     */
    WFDimensionPanelComponent.prototype.render;
}
