// @farris/kendo-binding/esm2015/lib/directives/primeng-tree-grid/FarrisPrimengTreeTableBindingDirective.js
/*******************************************************************************************************************
 *                                     FarrisPrimengTreeTableBindingDirective指令说明                               *
 * *****************************************************************************************************************
 *
 * 1. 依赖primeng版本: primeng:^9.0.0,primeicons:^4.0.0
 *
 * 2. 由于可编辑树表无法触发编辑单元格的关闭事件onEditComplete，因此对priemeng的node_modules\primeng\fesm5\primeng-treetable.js脚本作出修改：
 *    源码修改如下，+表示新增加的脚本语句:
 *
 *  TreeTable.prototype.updateEditingCell = function (cell, data, field) { //+ 参数增加data和field
 *      this.editingCell = cell;
 *      this.editingCellData = data; //+
 *      this.editingCellField = field; //+
 *      this.bindDocumentEditListener();
 *  };
 *
 *  TreeTable.prototype.bindDocumentEditListener = function () {
 *       var _this = this;
 *       if (!this.documentEditListener) {
 *           this.documentEditListener = function (event) {
 *               if (_this.editingCell && !_this.editingCellClick && _this.isEditingCellValid()) {
 *                   domhandler_1.DomHandler.removeClass(_this.editingCell, 'ui-editing-cell');
 *                   _this.editingCell = null;
 *                   _this.onEditComplete.emit({ field: _this.editingCellField, data: _this.editingCellData });//+
 *                   _this.unbindDocumentEditListener();
 *               }
 *               _this.editingCellClick = false;
 *           };
 *           document.addEventListener('click', this.documentEditListener);
 *       }
 *   };
 *
 *  TreeTable.prototype.bindDocumentEditListener = function () {
 *       var _this = this;
 *       if (!this.documentEditListener) {
 *           this.documentEditListener = function (event) {
 *               if (_this.editingCell && !_this.editingCellClick && _this.isEditingCellValid()) {
 *                   domhandler_1.DomHandler.removeClass(_this.editingCell, 'ui-editing-cell');
 *                   _this.editingCell = null;
 *                   _this.onEditComplete.emit({ field: _this.editingCellField, data: _this.editingCellData });//+
 *                   _this.unbindDocumentEditListener();
 *               }
 *               _this.editingCellClick = false;
 *           };
 *           document.addEventListener('click', this.documentEditListener);
 *       }
 *   };
 *
 *  TTEditableColumn.prototype.openCell = function () {
 *       var _this = this;
 *        this.tt.updateEditingCell(this.el.nativeElement,this.data, this.field);//+
 *        domhandler_1.DomHandler.addClass(this.el.nativeElement, 'ui-editing-cell');
 *        this.tt.onEditInit.emit({ field: this.field, data: this.data });
 *        this.zone.runOutsideAngular(function () {
 *            setTimeout(function () {
 *                var focusable = domhandler_1.DomHandler.findSingle(_this.el.nativeElement, 'input, textarea');
 *                if (focusable) {
 *                     focusable.focus();
 *                 }
 *             }, 50);
 *        });
 *   };
 *
 *  TTEditableColumn.prototype.onClick = function (event) {
 *        if (this.isEnabled()) {
 *            this.tt.editingCellClick = true;
 *            if (this.tt.editingCell) {
 *                if (this.tt.editingCell !== this.el.nativeElement) {
 *                    if (!this.tt.isEditingCellValid()) {
 *                        return;
 *                    }
 *                    this.tt.onEditComplete.emit({ field: this.tt.oldField, data: this.tt.oldDdata });//+
 *                    DomHandler.removeClass(this.tt.editingCell, 'ui-editing-cell');
 *                    this.openCell();
 *                }
 *            }
 *            else {
 *                this.openCell();
 *            }
 *        }
 *  };
 *
 *  TTEditableColumn.prototype.openCell = function () {
 *         var _this = this;
 *         this.tt.updateEditingCell(this.el.nativeElement,this.data, this.field);
 *         DomHandler.addClass(this.el.nativeElement, 'ui-editing-cell');
 *         this.tt.onEditInit.emit({ field: this.field, data: this.data });
 *         this.tt.oldField= this.field; //+
 *         this.tt.oldDdata=this.data;  //+
 *         this.zone.runOutsideAngular(function () {
 *             setTimeout(function () {
 *                 var focusable = DomHandler.findSingle(_this.el.nativeElement, 'input, textarea');
 *                if (focusable) {
 *                     focusable.focus();
 *                 }
 *           }, 50);
 *       });
 *   };
 *
 * 3. HTML模板
 * (1) 不启用多表头
 *      <p-treeTable [columns]="treeGridColumns" farrisPrimengTreeTableBinding farrisHierarchyInfoKey="layerField" [resizableColumns]="true" farrisPrimengTreeTableIdField="id">
 *             <ng-template pTemplate="header" let-columns>
 *                 <tr>
 *                     <th *ngFor="let col of columns" ttResizableColumn>
 *                         {{col.header}}
 *                     </th>
 *                 </tr>
 *             </ng-template>
 *             <ng-template pTemplate="body" let-rowNode let-rowData="rowData" let-columns="columns">
 *                 <tr>
 *                     <td *ngFor="let col of columns; let i = index" ttEditableColumn [ttEditableColumn]="rowData"
 *                         [ttEditableColumnField]="col.field">
 *                         <p-treeTableToggler [rowNode]="rowNode" *ngIf="i === 0"></p-treeTableToggler>
 *                         <p-treeTableCellEditor>
 *                             <ng-template pTemplate="input">
 *                                  <!-- 控件动态生成区域，此处可放置控件动态生成组件，生成规则如下：
 *                                            1. 字符串类型：文本控件
 *                                            2. 数字类型：数字控件
 *                                            3. 布尔类型：复选框控件
 *                                            4. 日期时间类型：日期时间控件
 *                                            5. 枚举类型：下拉框控件
 *                                            6. 帮助类型：帮助控件-->
 *                             </ng-template>
 *                             <ng-template pTemplate="output">{{ col | formatCellData : rowData }}</ng-template>
 *                         </p-treeTableCellEditor>
 *                     </td>
 *                 </tr>
 *             </ng-template>
 *     </p-treeTable>
 *
 * (2) 启用多表头
 *
 *  <p-treeTable [columns]="treeGridColumns" farrisPrimengTreeTableBinding farrisHierarchyInfoKey="layerField" [resizableColumns]="true" farrisPrimengTreeTableIdField="id">
 *             <ng-template pTemplate="header" let-columns>
 *                  <tr *ngFor="let columnHeader of treeGridColumnHeaders">
 *                      <th *ngFor="let col of columnHeader" ttResizableColumn [attr.rowspan]="col.rowspan" [attr.colspan]="col.colspan">
 *                          {{col.header}}
 *                      </th>
 *                  </tr>
 *             </ng-template>
 *             <ng-template pTemplate="body" let-rowNode let-rowData="rowData" let-columns="columns">
 *                 <tr>
 *                     <td *ngFor="let col of columns; let i = index" ttEditableColumn [ttEditableColumn]="rowData"
 *                         [ttEditableColumnField]="col.field">
 *                         <p-treeTableToggler [rowNode]="rowNode" *ngIf="i === 0"></p-treeTableToggler>
 *                         <p-treeTableCellEditor>
 *                             <ng-template pTemplate="input">
 *                                  <!-- 控件动态生成区域，此处可放置控件动态生成组件，生成规则如下：
 *                                            1. 字符串类型：文本控件
 *                                            2. 数字类型：数字控件
 *                                            3. 布尔类型：复选框控件
 *                                            4. 日期时间类型：日期时间控件
 *                                            5. 枚举类型：下拉框控件
 *                                            6. 帮助类型：帮助控件-->
 *                             </ng-template>
 *                             <ng-template pTemplate="output">{{ col | formatCellData : rowData }}</ng-template>
 *                         </p-treeTableCellEditor>
 *                     </td>
 *                 </tr>
 *             </ng-template>
 *  </p-treeTable>
 *
 * 多表头数据格式：
 *  this.treeGridColumnHeaders=[
 *     [{ header: 'Brand',rowspan:3, colspan:1},{ header: 'Sale Rate',rowspan:1, colspan:4}],
 *     [{ header: 'Sales',rowspan:1, colspan:2},{ header: 'Profits',rowspan:1, colspan:2}],
 *     [{ header: 'Last Year',rowspan:1, colspan:1},{ header: 'This Year',rowspan:1, colspan:1},{ header: 'Last Year',rowspan:1, colspan:1},{ header: 'This Year',rowspan:1, colspan:1}],
 *  ]

 **********************************************************************************************************************/
import { Directive, Input, HostListener, Output, EventEmitter } from '@angular/core';
import { FrameContext, ChangeType, ComponentType } from '@farris/devkit';
import { TreeTable } from 'primeng/treetable';
import { TreeNodeBuilderFactory } from '../tree_table_node_builder/tree_node_builder_factory';
import { DateTimeHelperService } from '@farris/ui-common/date';
const PARENT_ELEMENT = "parentElement";
const PATH = "path";
const IS_DETAIL = 'isDetail';
export class FarrisPrimengTreeTableBindingDirective {
    constructor(treeTable, frameContext, dateService) {
        this.treeTable = treeTable;
        this.frameContext = frameContext;
        this.dateService = dateService;
        this.selectedRowId = null;
        this.treeTableSelectedEvent = null;
        /**
         * 节点选择变更事件
         */
        this.nodeSelectChange = new EventEmitter();
        this.oldTreeNodeMap = {};
    }
    /**
     * 绑定数据
     */
    get bindingData() {
        return this.frameContext.bindingData;
    }
    /**
     * 绑定数据列表
     */
    get bindingList() {
        return this.bindingData.getList();
    }
    getHierarchyInfoField() {
        let infos = this.hierarchyInfoKey.split('/');
        return infos[infos.length - 1];
    }
    ngOnChanges(changes) {
    }
    ngOnInit() {
        //解析获取分级码字段编号
        this.hierarchyInfoField = this.getHierarchyInfoField();
        // 将分级码信息保存到root-component
        const virtualRootFrameContext = this.frameContext.getVirtualRootFrameContext();
        virtualRootFrameContext.setParam('hierarchyInfoKey', this.hierarchyInfoKey);
        this.bindingData.changes.subscribe((change) => {
            this.bindData(change);
            this.updateSelectedRow(change);
        });
        this.frameContext.viewModel.uiState.setPropertyValue('ids', []);
        if (this.frameContext && this.frameContext.frameComponent) {
            this.frameContext.frameComponent.componentType = ComponentType.primengTreeComponent;
        }
    }
    /**
     * 获取树列表绑定数据对应的实体数组
     * @returns 实体数组
     */
    getEntityArr() {
        const entityCollection = this.frameContext.repository.entityCollection;
        const bindingPath = this.frameContext.bindingData.bindingPath;
        if (!bindingPath || bindingPath === '/') {
            return entityCollection.toArray();
        }
        let bindingList = this.frameContext.bindingData.list;
        let bindingPathArr = bindingPath.split('/').filter(str => str);
        let pathArray = [];
        bindingPathArr.forEach((path) => {
            pathArray.push(`${bindingList.primaryKey}:${bindingList.currentId}`);
            pathArray.push(path);
            bindingList = bindingList.currentItem[path];
        });
        return entityCollection.getEntitiesByPath(pathArray).toArray();
    }
    /**
     * 将真实顺序标记到对应的实体上
     * @param entityArr 实体数组
     * @param treeNodes 树节点
     */
    setActualIndex(entityArr, treeNodes) {
        if (entityArr.length < 1)
            return;
        // 初始化序号器并依次遍历每一棵树
        let counter = { index: 0 };
        const primaryKey = entityArr[0].primaryKey;
        treeNodes.forEach((rootNode) => {
            this.traverseTree(rootNode, counter, entityArr, primaryKey);
        });
    }
    /**
     * 前序递归遍历树节点并在对应实体上标记序号
     * @param treeNode 树节点
     * @param counter 序号器
     * @param entityArr 实体数组
     * @description
     * 在进行前端校验时，按照Entity在数组中的下标生成（错误提示的）行号的方式不适用于树列表，
     * 因为树列表可以在已经添加的节点上加入子节点，而Entity总是在最后追加新对象，导致二者的顺序不一定相同，
     * 所以在绑定数据的时候将树节点在列表中的真实顺序通过一个附加的属性（__ACTUAL_INDEX__）赋给Entity示例，
     * 当Entity调用validate方法时，检查是否存在该属性，如果存在则应用该属性的值作为行号。
     */
    traverseTree(treeNode, counter, entityArr, primaryKey) {
        const INDEX_LABEL = "__ACTUAL_INDEX__";
        // 处理当前节点
        const entity = entityArr.find(e => e[primaryKey] === treeNode.data[primaryKey]);
        if (entity) {
            Object.defineProperty(entity, INDEX_LABEL, {
                value: counter.index
            });
            counter.index++;
        }
        // 递归遍历分支
        treeNode.children.forEach((rootNode) => {
            this.traverseTree(rootNode, counter, entityArr, primaryKey);
        });
    }
    /**
     * 处理数据变更
     * @param change 变更
     * @returns 是否跳过绑定数据的整体刷新
     * @description
     * 目的是尽量少的重新构造整个树列表，比如在发生值变化事件时尽量只更改变化的值就结束，
     * 如果任何变化都重新构造树，会导致Tab键切换功能难以实现等问题。
     */
    handleChange(change) {
        // 没有变更时，继续绑定数据刷新
        if (!change)
            return false;
        // 分类处理具体的变更
        switch (change.type) {
            case ChangeType.SelectionChanged:
                return this.handleSelectionChanged(change);
            case ChangeType.GlobalSelectionChanged:
                return this.handleSelectionChanged(change);
            case ChangeType.ValueChanged:
                return this.handleValueChanged(change);
            default:
                return false;
        }
    }
    /**
     * 处理行切换变更
     * @param change 变更
     * @returns 是否跳过绑定数据的整体刷新
     */
    handleSelectionChanged(change) {
        // 当前绑定路径下发生行切换事件，不进行数据刷新（为了支持Tab键切换单元格）
        if (change.type === ChangeType.SelectionChanged || change.type === ChangeType.GlobalSelectionChanged) {
            let path = change.path;
            let frmPath = this.bindingData.bindingPath.split("/").filter(str => str);
            if (path.length === frmPath.length) {
                let same = true;
                for (let i = 0; i < path.length; i++) {
                    if (path[i] !== frmPath[i])
                        same = false;
                }
                if (same)
                    return true;
            }
        }
        return false;
    }
    /**
     * 处理行切换变更
     * @param change 变更
     * @returns 是否跳过绑定数据的整体刷新
     */
    handleValueChanged(change) {
        if (change.type !== ChangeType.ValueChanged) {
            return false;
        }
        // 接收到值变化事件，尝试仅进行局部数据的更新（为了支持Tab键切换单元格）
        let path = change.path;
        let bindingPathArr = this.bindingData.bindingPath.split("/").filter(str => str);
        // 判断绑定路径是否相同，如果不相同则整体刷新
        for (let item of bindingPathArr) {
            if (path.length > 0) {
                if (item === path[0]) {
                    path.shift();
                }
                else {
                    return false;
                }
            }
        }
        // 通过id找到待更新的树节点
        let id = change.id;
        if (id) {
            const primaryKey = this.bindingList.primaryKey;
            let nodes = this.treeTable.value;
            let targetNode = null;
            for (let node of nodes) {
                let result = this.findTreeNodeById(node, primaryKey, id);
                if (result) {
                    targetNode = result;
                    break;
                }
            }
            if (targetNode && path.length === 1) {
                const propertyName = path.shift();
                if (targetNode.data.hasOwnProperty(propertyName) && change.hasOwnProperty("value")) {
                    targetNode.data[propertyName] = change.value;
                    return true;
                }
            }
        }
        return false;
    }
    /**
     * 通过主键值查找树节点
     * @param treeNode 树根节点
     * @param primaryKey 主键名
     * @param keyValue 主键值
     * @returns 树节点
     * @todo 递归遍历实现比较低效，考虑通过oldTreeNodeMap查找树节点
     */
    findTreeNodeById(treeNode, primaryKey, keyValue) {
        if (!treeNode)
            return null;
        if (treeNode.data.hasOwnProperty(primaryKey) && treeNode.data[primaryKey] === keyValue)
            return treeNode;
        if (treeNode.children) {
            for (let node of treeNode.children) {
                let result = this.findTreeNodeById(node, primaryKey, keyValue);
                if (result)
                    return result;
            }
        }
        return null;
    }
    /**
    * 绑定数据
    */
    bindData(change) {
        let skipRefresh = this.handleChange(change);
        if (skipRefresh) {
            return;
        }
        // this.sourceData = this.bindingList.toJSON({ ignoreMultiLangInput: true });
        const oldTreeNodes = this.treeTable.value;
        const treeNodesData = this.bindingList.toJSON({ ignoreMultiLangInput: true });
        let treeNodes = [];
        if (treeNodesData && treeNodesData.length > 0) {
            let hierarchyInfo = treeNodesData[0][this.hierarchyInfoField];
            const builder = TreeNodeBuilderFactory.getBuilder(hierarchyInfo);
            // 服务器返回的分级节点信息匹配不到现有工厂里面的任何一种情况。确认新加的builder已经在工厂中注册
            if (builder == null) {
                throw new Error('TreeTable builder is null!');
            }
            let expandLevel = -1;
            const treeExpandLevel = this.treeTable['expandLevel'];
            if (typeof treeExpandLevel !== 'undefined') {
                expandLevel = treeExpandLevel;
            }
            const context = {
                expandLevel,
                component: this.treeTable,
            };
            const treeNodesBuilder = new builder(treeNodesData, oldTreeNodes, this.bindingList.primaryKey, this.hierarchyInfoField, context);
            treeNodes = treeNodesBuilder.build();
            // 获取绑定数据对应的实体列表
            let entityArr = this.getEntityArr();
            // 根据treeNodes的顺序在对应的实体上添加序号字段__ACTUAL_INDEX__
            this.setActualIndex(entityArr, treeNodes);
        }
        if (!this.treeTable.lazy) {
            this.treeTable.totalRecords = (this.treeTable._value ? this.treeTable._value.length : 0);
            if (this.treeTable.sortMode == 'single' && this.treeTable.sortField)
                this.treeTable.sortSingle();
            else if (this.treeTable.sortMode == 'multiple' && this.treeTable.multiSortMeta)
                this.treeTable.sortMultiple();
            else if (this.treeTable.hasFilter()) //sort already filters
                this.treeTable._filter();
        }
        this.treeTable.tableService.onUIUpdate(treeNodes);
        this.treeTable.value = treeNodes;
        this.treeTable.updateSerializedValue();
        // this.treeNodeMap = {};
        // this.sourceDataClone = this.sourceData;
        // this.treeTable.value = [...this.treeNodesBuilder()];
        // this.treeTableValueHandler();
    }
    updateSelectedRow(change) {
        const currentId = this.bindingList.currentId;
        if (this.treeTableSelectedEvent) {
            this.treeTableSelectedEvent = false;
            this.selectedRowId = currentId;
            return;
        }
        const serializedValue = this.treeTable.serializedValue;
        const treeNode = serializedValue && serializedValue.find(item => item && item.node && item.node.data && item.node.data.id === currentId) || null;
        if (treeNode) {
            if (this.treeTable.isSelected(treeNode.node) || currentId === this.selectedRowId) {
                return;
            }
            this.selectedRowId = currentId;
            this.treeTable.handleRowClick({ rowNode: treeNode, originalEvent: { target: { nodeName: '' } } });
        }
    }
    onEditInitHandler(event) {
        this.nodeSelectChange.emit(event['data']);
        const id = event.data[this.idField];
        this.setCurrentIdToBindingList(id);
    }
    onEditCompleteHandler(event) {
        let fieldName = event.field;
        const column = this.columns.find(item => item.field === fieldName);
        if (fieldName && column && column.editor && column.editor.options && column.editor.options.type === 'EditorTypes.LOOKUP') {
            return;
        }
        if (event && event.hasOwnProperty('data')) {
            const primaryKeyValue = event.data[this.frameContext.repository.primaryKey];
            let value = this.getValue(fieldName, event.data);
            const bindingPath = this.frameContext.viewModel.bindingPath || '/';
            const bindingPaths = bindingPath.split('/').filter(p => p);
            const bindingList = this.bindingData.getValue(bindingPaths);
            const bindingObject = bindingList.findById(primaryKeyValue);
            const oldValue = this.getValue(fieldName, bindingObject);
            const column = this.treeTable.columns.find(item => item.field === fieldName);
            const dataType = column && column.dataType || null;
            if (dataType === 'date' || dataType === 'datetime') {
                if (this.dateService) {
                    const format = column && column.formatter && column.formatter.options && column.formatter.options.format || 'yyyy-MM-dd';
                    value = this.dateService.formatTo(value, format);
                    if (!value) {
                        value = null;
                    }
                }
            }
            else if (dataType === 'number') {
                value = Number(value);
            }
            if (value !== oldValue) {
                this.updateBindingList(primaryKeyValue, fieldName, value);
            }
        }
    }
    onNodeExpandHandler(event) {
        // this.setNodeExpandedValue(event);
        const selectedId = event && event.node && event.node.id || null;
        if (!selectedId) {
            return;
        }
        const virtualRootFrameContext = this.frameContext.getVirtualRootFrameContext();
        const expandNodes = virtualRootFrameContext.getParam('TREE_LATEST_EXPANDED_ID') || [];
        expandNodes.push(selectedId);
        virtualRootFrameContext.setParam('TREE_LATEST_EXPANDED_ID', expandNodes);
    }
    onNodeCollapseHandler(event) {
        // this.setNodeExpandedValue(event);
    }
    onNodeSelectHandler(event) {
        const id = event && event.node && event.node.data && event.node.data[this.frameContext.repository.primaryKey] || null;
        if (id === this.bindingList.currentId) {
            return;
        }
        this.treeTableSelectedEvent = true;
        this.setCurrentIdToBindingList(id);
    }
    onNodeUnselectHandler(event) {
        this.setCurrentIdToBindingList(null);
    }
    setNodeExpandedValue(event) {
        if (!event || !this.oldTreeNodeMap || Object.keys(this.oldTreeNodeMap).length <= 0) {
            return;
        }
        try {
            let currentId = event.node.data.id;
            this.oldTreeNodeMap[currentId].expanded = event.expanded;
        }
        catch (e) {
            console.warn(`setNodeExpandedValue error:${e.message}`);
        }
    }
    setCurrentIdToBindingList(selectedId) {
        const oldCurrentId = this.bindingList.currentId;
        if (oldCurrentId === selectedId) {
            return;
        }
        this.bindingList.setCurrentId(selectedId, true);
    }
    getPaths() {
        let paths = [];
        if (!this.frameContext.viewModel || !this.frameContext.viewModel.bindingPath || this.frameContext.viewModel.bindingPath === '/') {
            return paths;
        }
        else {
            let bindingPath = this.frameContext.viewModel.bindingPath;
            paths = bindingPath.substring(1, bindingPath.length).split('/');
            return paths;
        }
    }
    updateBindingList(primaryValue, propertyName, value) {
        if (!propertyName) {
            return;
        }
        // 更新主表部分行的字段
        const propertyNames = propertyName.split('.').filter(item => item);
        const bindingPath = this.getBindingPathArray();
        // 取出来的一定是bindingList
        const list = this.bindingData.getValue(bindingPath);
        // 修改的是当前行
        const paths = bindingPath.concat(propertyNames);
        if (list && primaryValue === list.currentItem.primaryKeyValue) {
            this.bindingData.setValue(paths, value, true, true);
            return;
        }
        const bindingObject = this.bindingList.findById(primaryValue);
        if (!bindingObject) {
            return;
        }
        if (propertyNames.length < 2) {
            bindingObject.setValue(propertyName, value, true, true, undefined, this.bindingData.getValudChangeInvokerFactory()(paths));
        }
        else {
            let targetBindingObject = null;
            const fpaths = propertyNames.slice(0, propertyNames.length - 1);
            const targetPropertyName = propertyNames[propertyNames.length - 1];
            fpaths.forEach(prop => {
                targetBindingObject = targetBindingObject && targetBindingObject[prop] || bindingObject[prop];
            });
            // todo:需要添加值变化事件
            targetBindingObject.setValue(targetPropertyName, value, true, true, undefined, this.bindingData.getValudChangeInvokerFactory()(paths));
        }
    }
    getBindingPathArray() {
        const path = this.frameContext.viewModel.bindingPath;
        if (path) {
            return path.split('/').filter(n => n !== '');
        }
        return [];
    }
    getValue(path, data) {
        let value = '';
        if (path.indexOf('.') === -1) {
            value = data[path];
        }
        else {
            value = path.split('.').reduce((result, prop) => {
                return result[prop];
            }, data);
        }
        return value;
    }
    setChecks(ids) {
        this.frameContext.viewModel.uiState.setPropertyValue('ids', ids);
    }
}
FarrisPrimengTreeTableBindingDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farrisPrimengTreeTableBinding]'
            },] }
];
/** @nocollapse */
FarrisPrimengTreeTableBindingDirective.ctorParameters = () => [
    { type: TreeTable },
    { type: FrameContext },
    { type: DateTimeHelperService }
];
FarrisPrimengTreeTableBindingDirective.propDecorators = {
    hierarchyInfoKey: [{ type: Input, args: ['farrisHierarchyInfoKey',] }],
    columns: [{ type: Input, args: ['columns',] }],
    idField: [{ type: Input, args: ['farrisPrimengTreeTableIdField',] }],
    nodeSelectChange: [{ type: Output }],
    onEditInitHandler: [{ type: HostListener, args: ['onEditInit', ['$event'],] }],
    onEditCompleteHandler: [{ type: HostListener, args: ['onEditComplete', ['$event'],] }],
    onNodeExpandHandler: [{ type: HostListener, args: ['onNodeExpand', ['$event'],] }],
    onNodeCollapseHandler: [{ type: HostListener, args: ['onNodeCollapse', ['$event'],] }],
    onNodeSelectHandler: [{ type: HostListener, args: ['onNodeSelect', ['$event'],] }],
    onNodeUnselectHandler: [{ type: HostListener, args: ['onNodeUnselect', ['$event'],] }]
};
