// @gsp-wf/wf-process-editor/esm2015/lib/views/performer/views/position-performer/position-performer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, ViewChild, ElementRef, NgZone, ComponentFactoryResolver, Injector, Output, EventEmitter, TemplateRef, ChangeDetectorRef, Optional, Inject, LOCALE_ID } from '@angular/core';
import { SysFarrisLookupComponent } from '@gsp-sys/sysmgr-ui';
import { ProcessDesignerUIState } from '../../../../app/process-designer.uistate';
import { ExprEditMode, ExpressionService, InjectedType } from '@gsp-svc/expression';
import { PositionPerformerEntity } from '../../entities/position-performer.entity';
import { SysLookupDetailComponent } from '@gsp-sys/sysmgr-lookup';
import { BsModalService } from '@farris/ui-modal';
import { WF_LANG_RESOURCES, WfLocalePipe } from '../../../../i18n/pipe/wf-process-editor-locale.pipe';
import { PositionsNamePipe } from './positions-name.pipe';
export class PositionPerformerComponent {
    /**
     * @param {?} designerUIState
     * @param {?} expr
     * @param {?} ngZone
     * @param {?} resolver
     * @param {?} injector
     * @param {?} modalService
     * @param {?} cdr
     * @param {?} localeId
     * @param {?} resources
     * @param {?} localePipe
     * @param {?} positionsNamePipe
     */
    constructor(designerUIState, expr, ngZone, resolver, injector, modalService, cdr, localeId, resources, localePipe, positionsNamePipe) {
        this.designerUIState = designerUIState;
        this.expr = expr;
        this.ngZone = ngZone;
        this.resolver = resolver;
        this.injector = injector;
        this.modalService = modalService;
        this.cdr = cdr;
        this.localeId = localeId;
        this.resources = resources;
        this.localePipe = localePipe;
        this.positionsNamePipe = positionsNamePipe;
        this.cls = 'd-flex flex-fill h-100';
        this.closeEmitter = new EventEmitter();
        this.submitEmitter = new EventEmitter();
        this.showClearButton = false;
        this.isRequired = false;
        this.groupIcon = '<i class="f-icon f-icon-lookup"></i>';
        this.positionTypes = [
            { value: '0', text: this.localePipe.transform('component.positionPerformer.orgPosition') },
            { value: '1', text: this.localePipe.transform('component.positionPerformer.generalPosition') }
        ];
        this.performerName = '';
        this.style = { background: '#FFF' };
        this.newFilter = {};
        this.treeFilter = {};
        this.curPosIdList = [];
        this.editState = true;
        this.showCondition = true;
        this.positions = [];
        this.localePipe = this.localePipe || new WfLocalePipe(localeId, resources);
        this.position = this.position || new PositionPerformerEntity();
        this.performerName = this.performerName || (this.positions && this.positions.length ? this.positionsNamePipe.transform(this.positions) : this.position.performerName || '');
        if (this.designerUIState.scene === 'inSuite') {
            this.position.positionType = '1';
            this.newFilter = { positionType: '1' };
            this.style = { background: '#FFF', height: '150px' };
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.designerUIState.getOrgConfig().subscribe((/**
         * @param {?} re
         * @return {?}
         */
        re => {
            if (re && re.currentValue === 'true') {
                this.isRequired = true;
            }
        }));
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        () => {
            this.inputGroup.nativeElement.addEventListener('mouseenter', this.onMouseEnter.bind(this));
            this.inputGroup.nativeElement.addEventListener('mouseleave', this.onMouseLeave.bind(this));
        }));
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onMouseEnter(event) {
        this.showClearButton = true;
        if (this.showClearButton) {
            this.toggleClearIcon(event, true);
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onMouseLeave(event) {
        this.showClearButton = false;
        this.toggleClearIcon(event, false);
    }
    /**
     * @private
     * @param {?} event
     * @param {?=} isShow
     * @return {?}
     */
    toggleClearIcon(event, isShow = false) {
        /** @type {?} */
        const str = isShow ? '' : 'none';
        /** @type {?} */
        const clearIcon = event.target.querySelector('.input-group-clear');
        if (clearIcon) {
            clearIcon.style.display = str;
        }
    }
    /**
     * @return {?}
     */
    clickOK() {
        this.submitEmitter.emit(this.position);
    }
    /**
     * @return {?}
     */
    clickCancel() {
        this.closeEmitter.emit();
    }
    /**
     * @return {?}
     */
    clear() {
        this.position.performerId = null;
        this.position.performerName = null;
        this.performerName = '';
    }
    /**
     * @param {?} data
     * @return {?}
     */
    positionTypeChange(data) {
        if (data.value === '0') {
            this.newFilter = { positionType: '0' };
            this.positionType = '0';
        }
        else {
            this.newFilter = { positionType: '1' };
            this.positionType = '1';
        }
        this.clear();
        this.position.orgType = null;
        this.position.orgId = null;
    }
    /**
     * @return {?}
     */
    selectOrganition() {
        // this.expr.clearContext();
        if (this.position.positionType === '1') {
            if (this.designerUIState.schemas.length > 0) {
                for (const schema of this.designerUIState.schemas) {
                    this.expr.addSchema('Schema', schema);
                }
                if (this.designerUIState.schemas[0].entityTypes && this.designerUIState.schemas[0].entityTypes.length > 0) {
                    /** @type {?} */
                    const entityType = this.designerUIState.schemas[0].entityTypes[0];
                    this.expr.addInjectedEntity(entityType.name, 'Schema', entityType.name, InjectedType.EntityType);
                }
            }
            /** @type {?} */
            const context = new Array();
            if (this.designerUIState.subActivityContext.length > 0) {
                for (const variable of this.designerUIState.subActivityContext) {
                    if (!context.some((/**
                     * @param {?} c
                     * @return {?}
                     */
                    c => c.key === variable.key))) {
                        context.push(variable);
                    }
                }
            }
            if (context.length > 0) {
                for (const v of context) {
                    this.expr.addContext(v);
                }
            }
            this.expr.setEditPattern(ExprEditMode.Advanced);
            this.expr.addExpressionText(this.position.orgId);
            this.expr.buildExpression().then((/**
             * @param {?} expr
             * @return {?}
             */
            (expr) => {
                this.position.orgId = expr;
                //点击确定后，去除表达式里面的新增的上下文变量
                if (context.length > 0) {
                    this.expr.exprEntity.contextEntities = this.expr.exprEntity.contextEntities.filter((/**
                     * @param {?} x
                     * @return {?}
                     */
                    x => !context.some((/**
                     * @param {?} y
                     * @return {?}
                     */
                    y => y.key === x.key))));
                }
            }));
        }
    }
    /**
     * @return {?}
     */
    selectPosition() {
        this.setPosition();
    }
    /**
     * @return {?}
     */
    editCondition() {
        // this.expr.clearContext();
        if (this.designerUIState.schemas.length > 0) {
            for (const schema of this.designerUIState.schemas) {
                this.expr.addSchema('Schema', schema);
            }
            if (this.designerUIState.schemas[0].entityTypes && this.designerUIState.schemas[0].entityTypes.length > 0) {
                /** @type {?} */
                const entityType = this.designerUIState.schemas[0].entityTypes[0];
                this.expr.addInjectedEntity(entityType.name, 'Schema', entityType.name, InjectedType.EntityType);
            }
        }
        /** @type {?} */
        let context = new Array();
        context = this.designerUIState.addContext(context);
        if (this.designerUIState.subActivityContext.length > 0) {
            for (const variable of this.designerUIState.subActivityContext) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                c => c.key === variable.key))) {
                    context.push(variable);
                }
            }
        }
        if (this.designerUIState.activityComponentContext.length > 0) {
            for (const variable of this.designerUIState.activityComponentContext) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                c => c.key === variable.key))) {
                    context.push(variable);
                }
            }
        }
        if (context.length > 0) {
            for (const v of context) {
                this.expr.addContext(v);
            }
        }
        this.expr.setEditPattern(ExprEditMode.Conditional); // 表达式展现模式：默认简易版
        this.expr.addExpressionText(this.position.condition);
        this.expr.buildExpression().then((/**
         * @param {?} expressioninfo
         * @return {?}
         */
        (expressioninfo) => {
            this.position.condition = expressioninfo;
            for (const p of this.positions) {
                p.condition = expressioninfo;
            }
            //点击确定后，去除表达式里面的新增的上下文变量
            if (context.length > 0) {
                this.expr.exprEntity.contextEntities = this.expr.exprEntity.contextEntities.filter((/**
                 * @param {?} x
                 * @return {?}
                 */
                x => !context.some((/**
                 * @param {?} y
                 * @return {?}
                 */
                y => y.key === x.key))));
            }
        }));
    }
    /**
     * @param {?} re
     * @return {?}
     */
    orgTypeChange(re) {
        this.position.orgType = re.data.value;
    }
    /**
     * @return {?}
     */
    clearOrgType() {
        this.position.orgType = null;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    positionsSelected(data) {
        this.positions = [];
        if (data || data.data.length > 0) {
            /** @type {?} */
            const positions = [];
            for (const dataSelected of data.data) {
                if (this.editState) {
                    this.position = new PositionPerformerEntity();
                    this.position.performerId = dataSelected.id;
                    this.position.performerName = dataSelected.name;
                    this.position.type = 'wf.PositionParticipantExpression';
                    if (!this.position.positionType) {
                        this.position.positionType = dataSelected.posType + '';
                        this.newFilter = {
                            positionType: dataSelected.posType
                        };
                    }
                    this.position.isRequired = this.isRequired;
                    this.position.condition = this.condition ? this.condition : "";
                    positions.push(this.position);
                }
                else {
                    this.position.performerId = dataSelected.id;
                    this.position.performerName = dataSelected.name;
                    this.position.type = 'wf.PositionParticipantExpression';
                    if (!this.position.positionType) {
                        this.position.positionType = dataSelected.posType + '';
                        this.newFilter = {
                            positionType: dataSelected.posType
                        };
                    }
                    this.position.isRequired = this.isRequired;
                    this.position.condition = this.condition ? this.condition : "";
                    positions.push(this.position);
                }
            }
            this.positions = [...positions];
            this.performerName = this.positions && this.positions.length ? (this.positionsNamePipe.transform(this.positions)) : (this.position.performerName ? this.position.performerName : '');
        }
    }
    /**
     * @return {?}
     */
    setPosition() {
        this.positions = [];
        /** @type {?} */
        const compFactory = this.resolver.resolveComponentFactory(SysLookupDetailComponent);
        /** @type {?} */
        const compRef = compFactory.create(this.injector);
        compRef.instance.initGridIdList = this.curPosIdList;
        compRef.instance.treeFilter = this.treeFilter;
        compRef.instance.listFilter = this.newFilter;
        compRef.instance.sourceType = 'position';
        compRef.instance.multiSelect = false;
        // 编辑状态下只能单选
        if (this.positionType === "0" && this.editState) {
            compRef.instance.multiSelect = true;
        }
        /** @type {?} */
        const modalConfig = {
            title: this.localePipe.transform('component.positionPerformer.sysPositions'),
            width: 1000,
            height: 550,
            showButtons: true,
            buttons: compRef.instance.btnTemplateRef,
            showMaxButton: true,
            resizable: false
        };
        /** @type {?} */
        const dialog = this.modalService.show(compRef, modalConfig);
        // 取消后
        compRef.instance.afterCancle.subscribe((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            // 消失模态框
            dialog.close();
        }));
        // 确认后
        compRef.instance.afterConfirm.subscribe((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            this.positionsSelected(res);
        }));
    }
    /**
     * @return {?}
     */
    getPerformers() {
        if (!this.editState) {
            this.positions.push(this.position);
        }
        return this.positions;
    }
}
PositionPerformerComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-position-performer',
                template: "<div class=\"d-flex flex-fill farris-overflow-y-auto p-md-2\" style=\"background: #FFF;\">\r\n  <div class=\"d-flex flex-fill farris-overflow-y-auto mx-3 my-2\" style=\"border-radius:16px;border: 1.5px solid #F1F3F7;\">\r\n  <form style=\"width: 100%;\">\r\n    <div class=\"d-flex m-2 flex-column w-100\" style=\"margin-top: 20px!important;\">\r\n      <div class=\"d-flex flex-column\" [ngStyle]=\"style\">\r\n        <div class=\"farris-form form-inline  farris-form-inline mr-5\">\r\n          <div *ngIf=\"designerUIState.scene!=='inSuite'\" class=\"col-12 farris-group-auto my-3\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{'component.positionPerformer.positionType'|wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <farris-combo-list [(ngModel)]=\"position.positionType\" name=\"positionType\" [panelHeight]=\"'auto'\" [idField]=\"'value'\"\r\n                    [textField]=\"'text'\" [data]=\"positionTypes\" (valueChange)=\"positionTypeChange($event)\"\r\n                    [panelHeight]=\"'auto'\" [enableClear]=\"false\">\r\n                  </farris-combo-list>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 farris-group-auto my-4\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-info text-danger\">*</span>\r\n                  <span class=\"farris-label-text\">{{designerUIState.scene==='inSuite'?('component.positionPerformer.role'|wfLocale):('component.positionPerformer.position'|wfLocale)}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <div #inputGroup class=\"input-group\">\r\n                    <input type=\"text\" class=\"form-control\" [(ngModel)]=\"performerName\" name=\"positionName\" [readOnly]=\"true\" />\r\n                    <div class=\"input-group-append\">\r\n                      <span [style.display]=\"showClearButton ? '' : 'none'\" class=\"input-group-text input-group-clear\" (click)=\"clear()\">\r\n                        <i class=\"f-icon modal_close\"></i>\r\n                      </span>\r\n                      <span class=\"input-group-text\" (click)=\"selectPosition()\" style=\"background: #FFF;\">\r\n                        <i class=\"f-icon f-icon-lookup\"></i>\r\n                      </span>\r\n                    </div>\r\n                  </div>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div *ngIf=\"designerUIState.scene!=='inSuite' && position.positionType === '1'\" class=\"col-12 farris-group-auto my-4\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <ng-container *ngIf=\"isRequired&&position.positionType==='1'\">\r\n                    <span class=\"farris-label-info text-danger\">*</span>\r\n                  </ng-container>\r\n                  <span class=\"farris-label-text\">{{'component.positionPerformer.sysType'|wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <farris-combo-list [(ngModel)]=\"position.orgType\" name=\"orgType\" [idField]=\"'value'\" [textField]=\"'text'\" [valueField]=\"'value'\"\r\n                    [data]=\"designerUIState.orgTypeCollection\" [panelHeight]=\"'auto'\" (clear)=\"clearOrgType()\"\r\n                    (selectChange)='orgTypeChange($event)' [readonly]=\"!(position.positionType==='1')\" placeholder=\"{{position.positionType==='1'?('component.positionPerformer.placeholder'|wfLocale):''}}\">\r\n                  </farris-combo-list>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div *ngIf=\"designerUIState.scene!=='inSuite' && position.positionType === '1'\" class=\"col-12 farris-group-auto my-4\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <ng-container *ngIf=\"isRequired&&position.positionType==='1'\">\r\n                    <span class=\"farris-label-info text-danger\">*</span>\r\n                  </ng-container>\r\n                  <span class=\"farris-label-text\">{{'component.positionPerformer.orgId'|wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <input-group class=\"form-control\" style=\"width: 100%;padding:0;border:0;\" [enableClear]=\"false\"\r\n                    [showButtonWhenDisabled]=\"false\" [readonly]=\"!(position.positionType==='1')\"\r\n                    [(ngModel)]=\"position.orgId\" name=\"organization\" placeholder=\"{{position.positionType==='1'?('component.positionPerformer.placeholder'|wfLocale):''}}\"\r\n                    [editable]=\"false\" (clickHandle)=\"selectOrganition()\" [groupText]=\"groupIcon\"></input-group>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <div *ngIf=\"showCondition\" class=\"d-flex flex-column\" style=\"background: #FFF;\">\r\n        <div class=\"farris-form form-inline farris-form-inline mr-5\">\r\n          <div class=\"col-12 farris-group-auto my-3\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{'component.positionPerformer.condition'|wfLocale}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <div class=\"input-group\">\r\n                    <input type=\"text\" class=\"form-control\" [(ngModel)]=\"this.position.condition\" name=\"conditon\" [readOnly]=\"true\" />\r\n                    <div class=\"input-group-append\">\r\n                      <span class=\"input-group-text\" (click)=\"editCondition()\" style=\"background: #FFF;\">\r\n                        <i class=\"f-icon f-icon-lookup\"></i>\r\n                      </span>\r\n                    </div>\r\n                  </div>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </form>\r\n</div>\r\n</div>\r\n\r\n<ng-template #buttonRef>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm px-3\" (click)=\"clickCancel()\">{{'component.positionPerformer.clickCancel'|wfLocale}}</button>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm px-3 mx-2\" (click)=\"clickOK()\">{{'component.positionPerformer.clickOK'|wfLocale}}</button>\r\n</ng-template>\r\n",
                providers: [
                    WfLocalePipe,
                ],
                styles: [""]
            }] }
];
/** @nocollapse */
PositionPerformerComponent.ctorParameters = () => [
    { type: ProcessDesignerUIState },
    { type: ExpressionService },
    { type: NgZone },
    { type: ComponentFactoryResolver },
    { type: Injector },
    { type: BsModalService },
    { type: ChangeDetectorRef, decorators: [{ type: Optional }] },
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [WF_LANG_RESOURCES,] }] },
    { type: WfLocalePipe, decorators: [{ type: Optional }] },
    { type: PositionsNamePipe }
];
PositionPerformerComponent.propDecorators = {
    cls: [{ type: HostBinding, args: ['class',] }],
    closeEmitter: [{ type: Output }],
    submitEmitter: [{ type: Output }],
    posiHelp: [{ type: ViewChild, args: [SysFarrisLookupComponent,] }],
    buttonRef: [{ type: ViewChild, args: ['buttonRef',] }],
    inputGroup: [{ type: ViewChild, args: ['inputGroup',] }]
};
if (false) {
    /** @type {?} */
    PositionPerformerComponent.prototype.cls;
    /** @type {?} */
    PositionPerformerComponent.prototype.closeEmitter;
    /** @type {?} */
    PositionPerformerComponent.prototype.submitEmitter;
    /** @type {?} */
    PositionPerformerComponent.prototype.posiHelp;
    /** @type {?} */
    PositionPerformerComponent.prototype.buttonRef;
    /** @type {?} */
    PositionPerformerComponent.prototype.inputGroup;
    /** @type {?} */
    PositionPerformerComponent.prototype.showClearButton;
    /** @type {?} */
    PositionPerformerComponent.prototype.isRequired;
    /** @type {?} */
    PositionPerformerComponent.prototype.groupIcon;
    /** @type {?} */
    PositionPerformerComponent.prototype.positionTypes;
    /** @type {?} */
    PositionPerformerComponent.prototype.performerName;
    /** @type {?} */
    PositionPerformerComponent.prototype.style;
    /** @type {?} */
    PositionPerformerComponent.prototype.condition;
    /** @type {?} */
    PositionPerformerComponent.prototype.newFilter;
    /** @type {?} */
    PositionPerformerComponent.prototype.treeFilter;
    /** @type {?} */
    PositionPerformerComponent.prototype.curPosIdList;
    /** @type {?} */
    PositionPerformerComponent.prototype.editState;
    /** @type {?} */
    PositionPerformerComponent.prototype.showCondition;
    /** @type {?} */
    PositionPerformerComponent.prototype.positionType;
    /** @type {?} */
    PositionPerformerComponent.prototype.position;
    /** @type {?} */
    PositionPerformerComponent.prototype.positions;
    /** @type {?} */
    PositionPerformerComponent.prototype.designerUIState;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.expr;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.resources;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.localePipe;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.positionsNamePipe;
}
