// @gsp-dip/data-imp-exp/esm2015/lib/design-time/dataie-rule-help/dataie-rule-help.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ComponentFactoryResolver, Injector } from '@angular/core';
import { Subject } from "rxjs";
import { BsModalService } from "@farris/ui-modal";
import { RuleIdHelpComponent } from "../rule-id-help/rule-id-help.component";
export class DataieRuleHelpComponent {
    /**
     * @param {?} injector
     * @param {?} modalService
     * @param {?} resolver
     */
    constructor(injector, modalService, resolver) {
        this.injector = injector;
        this.modalService = modalService;
        this.resolver = resolver;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * 打开导入导出规则id选择帮助
     * @param {?} ruleType 规则类型，导入为0，导出为1
     * @param {?} voId
     * @param {?} beId
     * @return {?}
     */
    showRuleIdHelp(ruleType, voId, beId) {
        /** @type {?} */
        const subject = new Subject();
        /** @type {?} */
        const compFactory = this.resolver.resolveComponentFactory(RuleIdHelpComponent);
        /** @type {?} */
        const compRef = compFactory.create(this.injector);
        compRef.instance.beId = beId;
        compRef.instance.ruleType = ruleType;
        compRef.instance.voId = voId;
        /** @type {?} */
        const modalConfig = {
            title: '规则选择',
            width: 920,
            height: 500,
            buttons: compRef.instance.modalFooter,
            showHeader: true,
            showMaxButton: false,
            showButtons: true
        };
        /** @type {?} */
        const dialog = this.modalService.show(compRef, modalConfig);
        compRef.instance.closeModal.subscribe((/**
         * @param {?} ruleId
         * @return {?}
         */
        ruleId => {
            console.log(ruleId);
            dialog.close();
            subject.next(ruleId);
        }));
        return subject;
    }
}
DataieRuleHelpComponent.decorators = [
    { type: Component, args: [{
                selector: 'dataie-rule-help',
                template: "\n",
                styles: [""]
            }] }
];
/** @nocollapse */
DataieRuleHelpComponent.ctorParameters = () => [
    { type: Injector },
    { type: BsModalService },
    { type: ComponentFactoryResolver }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataieRuleHelpComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    DataieRuleHelpComponent.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    DataieRuleHelpComponent.prototype.resolver;
}
