// @gsp-wf/wf-process-model/esm2015/lib/model/parameter/FormalParameter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelHelper, MexElement, DataType } from "@edp-pmf/bpmn-model";
import { ParameterMode } from "./ParameterMode";
export class FormalParameter extends MexElement {
    /**
     * @param {?} model
     * @param {?=} id
     * @param {?=} code
     * @param {?=} name
     * @param {?=} desc
     * @param {?=} index
     * @param {?=} datatype
     * @param {?=} mode
     */
    constructor(model, id, code, name, desc, index, datatype, mode) {
        super(model);
        this.Id = id || '';
        this.Code = code || '';
        this.Name = name || '';
        this.Description = desc || '';
        this.index = index || 0;
        this.DataType = datatype || new DataType(model);
        this.Mode = mode || ParameterMode.IN;
    }
    /**
     * @return {?}
     */
    get Index() {
        return this.index;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    set Index(index) {
        this.index = index;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "Id", this.Id);
        BpmnModelHelper.SetAttributeValue(obj, "Code", this.Code);
        BpmnModelHelper.SetAttributeValue(obj, "Name", this.Name);
        BpmnModelHelper.SetAttributeValue(obj, "Description", this.Description);
        BpmnModelHelper.SetAttributeValue(obj, "index", this.index);
        BpmnModelHelper.SetAttributeValue(obj, "DataType", this.DataType.ConvertToJsonObject());
        BpmnModelHelper.SetAttributeValue(obj, "Mode", this.Mode);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.Id = obj.Id;
        this.Name = obj.Name;
        this.index = obj.index;
        this.Description = obj.Description;
        this.Code = obj.Code;
        /** @type {?} */
        let type = new DataType(this.Model);
        type.LoadFromJsonObject(obj.DataType);
        this.DataType = type;
        this.Mode = ((/** @type {?} */ (obj.Mode)));
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const f = new FormalParameter(this.Model);
        return f;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    FormalParameter.prototype.index;
    /** @type {?} */
    FormalParameter.prototype.Id;
    /** @type {?} */
    FormalParameter.prototype.Code;
    /** @type {?} */
    FormalParameter.prototype.Name;
    /** @type {?} */
    FormalParameter.prototype.Description;
    /** @type {?} */
    FormalParameter.prototype.DataType;
    /** @type {?} */
    FormalParameter.prototype.Mode;
}
