// @gsp-wf/wf-process-model/esm2015/lib/model/parameter/WFActualParameter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModel, BpmnModelHelper, MexExtensionElement } from "@edp-pmf/bpmn-model";
export class ActualParameter extends MexExtensionElement {
    /**
     * @param {?} model
     * @param {?=} name
     * @param {?=} value
     * @param {?=} valueText
     * @param {?=} tag
     */
    constructor(model, name, value, valueText, tag) {
        super(model);
        this.ClrTypeID = ActualParameter.Clr_Type_ID;
        this.Name = name;
        this.Value = value;
        this.ValueText = valueText;
        this.Tag = tag;
    }
    /**
     * @return {?}
     */
    NewInstance() {
        return new ActualParameter(this.Model);
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelHelper.SetAttributeValue(obj, "Name", this.Name);
        BpmnModelHelper.SetAttributeValue(obj, "Value", this.Value);
        BpmnModelHelper.SetAttributeValue(obj, "ValueText", this.ValueText);
        BpmnModelHelper.SetAttributeValue(obj, "Tag", this.Tag);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.Name = obj.Name;
        this.Value = obj.Value;
        this.ValueText = obj.ValueText;
        this.Tag = obj.Tag;
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const parameter = new ActualParameter(this.Model);
        parameter.Name = this.Name;
        parameter.Value = this.Value;
        parameter.Tag = this.Tag;
        parameter.ValueText = this.ValueText;
        return parameter;
    }
}
ActualParameter.Clr_Type_ID = "wf.Mex.ActualParameter";
if (false) {
    /** @type {?} */
    ActualParameter.Clr_Type_ID;
    /** @type {?} */
    ActualParameter.prototype.ClrTypeID;
    /** @type {?} */
    ActualParameter.prototype.Name;
    /** @type {?} */
    ActualParameter.prototype.Value;
    /** @type {?} */
    ActualParameter.prototype.ValueText;
    /** @type {?} */
    ActualParameter.prototype.Tag;
}
BpmnModelHelper.RegisterElementType(new ActualParameter(new BpmnModel()));
