// @farris/ui-locale/esm2015/lib/locale.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonUtils } from '@farris/ui-common';
import { Injectable, LOCALE_ID, Inject, InjectionToken, Optional } from '@angular/core';
import { FARRIS_LOCALES } from './locales';
import * as i0 from "@angular/core";
import * as i1 from "@farris/ui-common";
/** @type {?} */
export const FARRIS_LOCAL_CUSTOM_DATA = new InjectionToken('自定义语言数据， 格式：{ "languageCode": { "name": { "key": "value" } } }');
export class LocaleService {
    /**
     * @param {?} localeId
     * @param {?} utils
     * @param {?} localeData
     */
    constructor(localeId, utils, localeData) {
        this.localeId = localeId;
        this.utils = utils;
        this.localeData = localeData;
        if (!localeId) {
            localeId = 'zh-CHS';
        }
        // 合并语言资源
        this.setLocaleData(localeData);
    }
    /**
     * 获取语言资源
     *  path : 资源路径，如:
     *  ```
     *  getResources('zh-CHS') // 返回所有中文语言资源
     *  getResources('zh-CHS.lookup') // 返回所有中文下帮助的资源
     *  ```
     * @param {?=} path
     * @return {?}
     */
    getResources(path = '') {
        if (path) {
            return this.utils.getValue(path, FARRIS_LOCALES);
        }
        return FARRIS_LOCALES;
    }
    /**
     * @param {?} ctrlName
     * @return {?}
     */
    getComponentOptions(ctrlName) {
        return this.langData[ctrlName];
    }
    /**
     * @param {?} propertyName
     * @return {?}
     */
    getValue(propertyName) {
        /** @type {?} */
        const val = this.utils.getValue(propertyName, this.langData);
        if (val) {
            return val;
        }
        else {
            // console.warn(`Can\'t find language Data for the ${propertyName}`);
            return '';
        }
    }
    /**
     * 合并现有的多语资源
     * @param {?=} localeData
     * @return {?}
     */
    setLocaleData(localeData = null) {
        if (localeData) {
            // 解析得到提供的语言编码
            /** @type {?} */
            const langCodes = Object.keys(localeData);
            langCodes.forEach((/**
             * @param {?} code
             * @return {?}
             */
            (code) => {
                /** @type {?} */
                const resName = Object.keys(localeData[code]);
                resName.forEach((/**
                 * @param {?} k
                 * @return {?}
                 */
                k => {
                    this.appendLanguageResource(k, localeData[code][k], code);
                }));
            }));
        }
        this.langData = FARRIS_LOCALES[this.localeId];
        if (!this.langData) {
            this.langData = FARRIS_LOCALES['zh-CHS'];
        }
    }
    /**
     * @private
     * @param {?} key
     * @param {?} data
     * @param {?=} lang
     * @return {?}
     */
    appendLanguageResource(key, data, lang = this.localeId) {
        if (!lang) {
            this.langData[key] = this.langData[key] || {};
            this.langData[key] = Object.assign(this.langData[key], data || {});
        }
        else {
            FARRIS_LOCALES[lang][key] = data || {};
        }
    }
}
LocaleService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
LocaleService.ctorParameters = () => [
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] },
    { type: CommonUtils },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [FARRIS_LOCAL_CUSTOM_DATA,] }] }
];
/** @nocollapse */ LocaleService.ngInjectableDef = i0.defineInjectable({ factory: function LocaleService_Factory() { return new LocaleService(i0.inject(i0.LOCALE_ID), i0.inject(i1.CommonUtils), i0.inject(FARRIS_LOCAL_CUSTOM_DATA, 8)); }, token: LocaleService, providedIn: "root" });
if (false) {
    /** @type {?} */
    LocaleService.prototype.langData;
    /** @type {?} */
    LocaleService.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    LocaleService.prototype.utils;
    /**
     * @type {?}
     * @private
     */
    LocaleService.prototype.localeData;
}
