// @farris/ui-locale/esm2015/lib/locale.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FarrisCommonModule } from '@farris/ui-common';
import { FarrisLocalePipe } from './locale.pipe';
import { NgModule } from '@angular/core';
import { LocaleService, FARRIS_LOCAL_CUSTOM_DATA } from './locale.service';
export class LocaleModule {
    /**
     * @param {?=} customData
     * @return {?}
     */
    static forRoot(customData) {
        return {
            ngModule: LocaleModule,
            providers: [
                { provide: FARRIS_LOCAL_CUSTOM_DATA, useValue: customData || {} },
                LocaleService
            ]
        };
    }
}
LocaleModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    FarrisLocalePipe
                ],
                imports: [
                    FarrisCommonModule.forRoot()
                ],
                exports: [FarrisLocalePipe],
                providers: [
                    LocaleService
                ]
            },] }
];
