// @gsp-wf/wf-process-model/esm5/lib/model/extension/condition-expression/SemanticExpression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModel, BpmnModelHelper } from "@edp-pmf/bpmn-model";
import { MexExpressionImpl } from "../../mex/MexExpressionImpl";
import { ConditionType } from "./ConditionType";
import { ConditionValueVersion } from "./ConditionValueVersion";
var SemanticExpression = /** @class */ (function (_super) {
    tslib_1.__extends(SemanticExpression, _super);
    function SemanticExpression(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = SemanticExpression.Clr_Type_ID;
        _this._value = '';
        _this.valueVersion = ConditionValueVersion.Unsure;
        _this.type = ConditionType.CONDITION;
        return _this;
    }
    /**
     * @return {?}
     */
    SemanticExpression.prototype.NewInstance = /**
     * @return {?}
     */
    function () {
        return new SemanticExpression(this.Model);
    };
    Object.defineProperty(SemanticExpression.prototype, "ConditionType", {
        get: /**
         * @return {?}
         */
        function () { return this.type; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this.type = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SemanticExpression.prototype, "Value", {
        get: /**
         * @return {?}
         */
        function () { return this._value; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this._value = value; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SemanticExpression.prototype, "ValueVersion", {
        get: /**
         * @return {?}
         */
        function () { return this.valueVersion; },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) { this.valueVersion = value; },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    SemanticExpression.prototype.ToString = /**
     * @return {?}
     */
    function () {
        return "Condition[value='" + this._value + "'" + ", type=" + this.type + "]";
    };
    /**
     * @protected
     * @return {?}
     */
    SemanticExpression.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Value = '';
    };
    /**
     * @return {?}
     */
    SemanticExpression.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "ConditionType", this.ConditionType);
        BpmnModelHelper.SetAttributeValue(obj, "Value", this.Value);
        BpmnModelHelper.SetAttributeValue(obj, "ValueVersion", this.ValueVersion);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    SemanticExpression.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.ConditionType = obj.ConditionType;
        this.Value = obj.Value;
        this.ValueVersion = obj.ValueVersion;
    };
    /**
     * @return {?}
     */
    SemanticExpression.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var condition = new SemanticExpression(this.Model);
        condition.Value = this.Value;
        condition.ValueVersion = this.valueVersion;
        condition.ConditionType = this.ConditionType;
        return condition;
    };
    SemanticExpression.Clr_Type_ID = 'wf.SemanticExpression';
    return SemanticExpression;
}(MexExpressionImpl));
export { SemanticExpression };
if (false) {
    /** @type {?} */
    SemanticExpression.Clr_Type_ID;
    /** @type {?} */
    SemanticExpression.prototype.ClrTypeID;
    /**
     * @type {?}
     * @private
     */
    SemanticExpression.prototype._value;
    /**
     * @type {?}
     * @private
     */
    SemanticExpression.prototype.valueVersion;
    /**
     * @type {?}
     * @private
     */
    SemanticExpression.prototype.type;
}
BpmnModelHelper.RegisterElementType(new SemanticExpression(new BpmnModel()));
