// @farris/ui-datagrid/esm2015/lib/utils/safe.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-12 10:11:59
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-08-12 10:12:05
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Pipe } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
export class SafePipe {
    /**
     * @param {?} sanitizer
     */
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
        this.dangerWords = ['<script', 'onerror', 'alert('];
    }
    /**
     * @param {?} value
     * @param {?} type
     * @return {?}
     */
    transform(value, type) {
        switch (type) {
            case 'html': {
                if (value === '' || value === null || value === undefined) {
                    return value;
                }
                // if (!this.hasDanger(value)) {
                //     return value;
                // }
                return this.sanitizer.bypassSecurityTrustHtml(value);
            }
            case 'style': return this.sanitizer.bypassSecurityTrustStyle(value);
            case 'script': return this.sanitizer.bypassSecurityTrustScript(value);
            case 'url': return this.sanitizer.bypassSecurityTrustUrl(value);
            case 'resourceUrl': return this.sanitizer.bypassSecurityTrustResourceUrl(value);
            default: throw new Error(`Invalid safe type specified: ${type}`);
        }
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    hasDanger(value) {
        if (typeof value === 'number' || typeof value === 'boolean') {
            return false;
        }
        /** @type {?} */
        let r = 0;
        this.dangerWords.forEach((/**
         * @param {?} n
         * @return {?}
         */
        n => {
            if (value.indexOf(n) > -1) {
                r++;
            }
        }));
        return r !== 0;
    }
}
SafePipe.decorators = [
    { type: Pipe, args: [{
                name: 'safe'
            },] }
];
/** @nocollapse */
SafePipe.ctorParameters = () => [
    { type: DomSanitizer }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    SafePipe.prototype.dangerWords;
    /**
     * @type {?}
     * @protected
     */
    SafePipe.prototype.sanitizer;
}
