// @edp-aif/runtime-api/esm2015/lib/aif-manual-creation/form-args/FormEntityData.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
export class FormEntityData extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        this.children = {};
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.dataId = null;
        if (this.children != null) {
            for (var key in this.children) {
                delete this.children[key];
            }
        }
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "dataId", this.dataId);
        if (this.children != null) {
            /** @type {?} */
            let childrenObj = {};
            for (let key in this.children) {
                BpmnModelJsonHelper.SetAttributeValue(childrenObj, key, this.children[key]);
            }
            BpmnModelJsonHelper.SetAttributeValue(obj, "children", childrenObj);
        }
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.dataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "dataId");
        /** @type {?} */
        let childrenObj = BpmnModelJsonHelper.GetAttributeValue(obj, "children");
        if (childrenObj != null) {
            for (let key in childrenObj) {
                /** @type {?} */
                let value = BpmnModelJsonHelper.GetAttributeValueAsList(childrenObj, key, FormEntityData);
                this.children[key] = value;
            }
        }
    }
}
if (false) {
    /** @type {?} */
    FormEntityData.prototype.dataId;
    /** @type {?} */
    FormEntityData.prototype.children;
}
