// @edp-aif/runtime-api/esm2015/lib/aif-manual-creation/form-args/FormCreationRuleExecutorArgs.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from '@edp-aif/common-api';
/**
 * 表单手工生单时执行生单规则参数类。
 */
export class FormCreationRuleExecutorArgs extends JsonConvertableImpl {
    constructor() {
        super(...arguments);
        /**
         * 来源单据ID树形列表
         */
        this.sourceVoEntityDatas = new Array();
        /**
         * 目标单据主表ID列表
         */
        this.targetVoDataIds = new Array();
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.sourceVoEntityDatas = new Array();
        this.creationRuleId = null;
        this.bizFlowchartId = null;
        this.targetVoDataIds = new Array();
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "bizFlowchartId", this.bizFlowchartId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "sourceVoEntityDatas", BpmnModelJsonHelper.StringListToJsonArray(this.sourceVoEntityDatas));
        BpmnModelJsonHelper.SetAttributeValue(obj, "targetVoDataIds", BpmnModelJsonHelper.StringListToJsonArray(this.targetVoDataIds));
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.bizFlowchartId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "bizFlowchartId");
        this.sourceVoEntityDatas = BpmnModelJsonHelper.StringListFromJsonArray(BpmnModelJsonHelper.GetAttributeValue(obj, "sourceVoEntityDatas"));
        this.targetVoDataIds = BpmnModelJsonHelper.StringListFromJsonArray((/** @type {?} */ (BpmnModelJsonHelper.GetAttributeValue(obj, "targetVoDataIds"))));
    }
}
if (false) {
    /**
     * 生单规则ID
     * @type {?}
     */
    FormCreationRuleExecutorArgs.prototype.creationRuleId;
    /**
     * 业务场景图ID
     * @type {?}
     */
    FormCreationRuleExecutorArgs.prototype.bizFlowchartId;
    /**
     * 来源单据ID树形列表
     * @type {?}
     */
    FormCreationRuleExecutorArgs.prototype.sourceVoEntityDatas;
    /**
     * 目标单据主表ID列表
     * @type {?}
     */
    FormCreationRuleExecutorArgs.prototype.targetVoDataIds;
}
