// @farris/ui-datagrid/esm5/lib/components/header/drag-drop-column.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var DragDropColumnService = /** @class */ (function () {
    function DragDropColumnService() {
    }
    /** 元素位置调整
     * arr: 数组
     * newIndex: 新位置索引
     * removedIndex: 老位置索引
     */
    /**
     * 元素位置调整
     * arr: 数组
     * newIndex: 新位置索引
     * removedIndex: 老位置索引
     * @param {?} arr
     * @param {?} newIndex
     * @param {?} removedIndex
     * @return {?}
     */
    DragDropColumnService.prototype.moveItem = /**
     * 元素位置调整
     * arr: 数组
     * newIndex: 新位置索引
     * removedIndex: 老位置索引
     * @param {?} arr
     * @param {?} newIndex
     * @param {?} removedIndex
     * @return {?}
     */
    function (arr, newIndex, removedIndex) {
        /** @type {?} */
        var removedItem = arr.splice(removedIndex, 1);
        arr.splice.apply(arr, tslib_1.__spread([newIndex, 0], removedItem));
        return arr;
    };
    /**
     * @param {?} columns
     * @return {?}
     */
    DragDropColumnService.prototype.backupColumns = /**
     * @param {?} columns
     * @return {?}
     */
    function (columns) {
        this.initColumns = columns.map((/**
         * @param {?} cols
         * @return {?}
         */
        function (cols) {
            return cols.map((/**
             * @param {?} n
             * @return {?}
             */
            function (n) {
                return {
                    id: n.id,
                    field: n.field
                };
            }));
        }));
    };
    DragDropColumnService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    DragDropColumnService.ctorParameters = function () { return []; };
    /** @nocollapse */ DragDropColumnService.ngInjectableDef = i0.defineInjectable({ factory: function DragDropColumnService_Factory() { return new DragDropColumnService(); }, token: DragDropColumnService, providedIn: "root" });
    return DragDropColumnService;
}());
export { DragDropColumnService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DragDropColumnService.prototype.initColumns;
}
