// @farris/ui-wizard/esm5/lib/page-detail/page-detail.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Directive, TemplateRef, ContentChild, HostBinding, Output, EventEmitter, Optional } from '@angular/core';
import { PageService } from '../service/page.service';
import { WizardService } from '../service/wizard.service';
import { WizardEventType } from '../model/wizard.model';
var wizardPageHeaderTemplateDirective = /** @class */ (function () {
    function wizardPageHeaderTemplateDirective(template) {
        this.template = template;
    }
    wizardPageHeaderTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[wizardPageHeader]'
                },] }
    ];
    /** @nocollapse */
    wizardPageHeaderTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return wizardPageHeaderTemplateDirective;
}());
export { wizardPageHeaderTemplateDirective };
if (false) {
    /** @type {?} */
    wizardPageHeaderTemplateDirective.prototype.template;
}
var wizardPageContentTemplateDirective = /** @class */ (function () {
    function wizardPageContentTemplateDirective(template) {
        this.template = template;
    }
    wizardPageContentTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[wizardPageContent]'
                },] }
    ];
    /** @nocollapse */
    wizardPageContentTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return wizardPageContentTemplateDirective;
}());
export { wizardPageContentTemplateDirective };
if (false) {
    /** @type {?} */
    wizardPageContentTemplateDirective.prototype.template;
}
var wizardPageFooterTemplateDirective = /** @class */ (function () {
    function wizardPageFooterTemplateDirective(template) {
        this.template = template;
    }
    wizardPageFooterTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[wizardPageFooter]'
                },] }
    ];
    /** @nocollapse */
    wizardPageFooterTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return wizardPageFooterTemplateDirective;
}());
export { wizardPageFooterTemplateDirective };
if (false) {
    /** @type {?} */
    wizardPageFooterTemplateDirective.prototype.template;
}
var FarrisPageDetailComponent = /** @class */ (function () {
    function FarrisPageDetailComponent(wizardSer, pageSer) {
        var _this = this;
        this.wizardSer = wizardSer;
        this.pageSer = pageSer;
        this.cls = true;
        this.pageId = '';
        this.pagedetailSelected = new EventEmitter();
        if (!this.pageSer) {
            this.pageSer = new PageService();
        }
        this.pageSer.activeState.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            //console.log('FarrisPageDetailComponent——' + this.pageId+'——'+data);
            if (data.state) {
                _this.pagedetailSelected.emit(data);
            }
        }));
    }
    Object.defineProperty(FarrisPageDetailComponent.prototype, "fill", {
        get: /**
         * @return {?}
         */
        function () { return this.wizardSer.getPageFill(); },
        enumerable: true,
        configurable: true
    });
    ;
    /**
     * @return {?}
     */
    FarrisPageDetailComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.pageId = this.pageSer.getPageId();
    };
    /**
     * @return {?}
     */
    FarrisPageDetailComponent.prototype.getPageId = /**
     * @return {?}
     */
    function () {
        return this.pageId;
    };
    /**
     * @return {?}
     */
    FarrisPageDetailComponent.prototype.getPrevPageId = /**
     * @return {?}
     */
    function () {
        return this.wizardSer.getPrePageId(this.pageId);
    };
    /**
     * @return {?}
     */
    FarrisPageDetailComponent.prototype.getWizardData = /**
     * @return {?}
     */
    function () {
        return this.wizardSer.getWizardData();
    };
    /**
     * 获取某页数据
     */
    /**
     * 获取某页数据
     * @param {?=} pageid
     * @return {?}
     */
    FarrisPageDetailComponent.prototype.getPageData = /**
     * 获取某页数据
     * @param {?=} pageid
     * @return {?}
     */
    function (pageid) {
        if (pageid === void 0) { pageid = this.pageId; }
        return this.wizardSer.getPageData(pageid);
    };
    /**
     * 更新某页数据
     * @param data
     */
    /**
     * 更新某页数据
     * @param {?} data
     * @param {?=} pageid
     * @return {?}
     */
    FarrisPageDetailComponent.prototype.setPageData = /**
     * 更新某页数据
     * @param {?} data
     * @param {?=} pageid
     * @return {?}
     */
    function (data, pageid) {
        if (pageid === void 0) { pageid = this.pageId; }
        this.wizardSer.setPageData(data, pageid);
    };
    /**
     * 下一步
     */
    /**
     * 下一步
     * @return {?}
     */
    FarrisPageDetailComponent.prototype.nextStep = /**
     * 下一步
     * @return {?}
     */
    function () {
        this.wizardSer.updateWizardState(WizardEventType.Next, this.pageId);
    };
    /**
     * 上一步
     */
    /**
     * 上一步
     * @return {?}
     */
    FarrisPageDetailComponent.prototype.prevStep = /**
     * 上一步
     * @return {?}
     */
    function () {
        this.wizardSer.updateWizardState(WizardEventType.Prev, this.pageId);
    };
    /**
     * 结束向导
     */
    /**
     * 结束向导
     * @return {?}
     */
    FarrisPageDetailComponent.prototype.finishWizard = /**
     * 结束向导
     * @return {?}
     */
    function () {
        this.wizardSer.updateWizardState(WizardEventType.Finish, this.pageId);
    };
    /**
     * 取消向导
     */
    /**
     * 取消向导
     * @return {?}
     */
    FarrisPageDetailComponent.prototype.cancelWizard = /**
     * 取消向导
     * @return {?}
     */
    function () {
        this.wizardSer.updateWizardState(WizardEventType.Cancel, this.pageId);
    };
    FarrisPageDetailComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wizard-page-detail',
                    template: "<div class=\"f-component-wizard-page-detail-header\" *ngIf=\"pageHeaderTemplate\">\r\n    <ng-container [ngTemplateOutlet]=\"pageHeaderTemplate\"></ng-container>\r\n</div>\r\n<!-- <div class=\"page-message\">\r\n                <ng-container [ngTemplateOutlet]=\"pageMessageTemplate\"></ng-container>\r\n        </div> -->\r\n<div class=\"f-component-wizard-page-detail-content\" *ngIf=\"pageContentTemplate\">\r\n    <ng-container [ngTemplateOutlet]=\"pageContentTemplate\"></ng-container>\r\n</div>\r\n<div class=\"f-component-wizard-page-detail-content\" *ngIf=\"!pageContentTemplate\">\r\n    <ng-content></ng-content>\r\n</div>\r\n<div class=\"f-component-wizard-page-detail-footer\" *ngIf=\"pageFooterTemplate\">\r\n    <ng-container [ngTemplateOutlet]=\"pageFooterTemplate\"></ng-container>\r\n</div>\r\n",
                    styles: [":host.f-component-wizard-page-detail-fill{-webkit-box-flex:1;flex-grow:1;flex-shrink:1;flex-basis:0;display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}:host.f-component-wizard-page-detail-fill .f-component-wizard-page-detail-footer,:host.f-component-wizard-page-detail-fill .f-component-wizard-page-detail-headeer{flex-shrink:0}:host.f-component-wizard-page-detail-fill .f-component-wizard-page-detail-content{-webkit-box-flex:1;flex-grow:1;flex-shrink:1;flex-basis:0;overflow:auto}:host.f-component-wizard-page-detail-fill .f-component-wizard-page-detail-footer{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;-webkit-box-align:center;align-items:center;-webkit-box-pack:end;justify-content:flex-end}"]
                }] }
    ];
    /** @nocollapse */
    FarrisPageDetailComponent.ctorParameters = function () { return [
        { type: WizardService },
        { type: PageService, decorators: [{ type: Optional }] }
    ]; };
    FarrisPageDetailComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class.f-component-wizard-page-detail',] }],
        fill: [{ type: HostBinding, args: ['class.f-component-wizard-page-detail-fill',] }],
        pageHeaderTemplate: [{ type: ContentChild, args: [wizardPageHeaderTemplateDirective, { read: TemplateRef },] }],
        pageContentTemplate: [{ type: ContentChild, args: [wizardPageContentTemplateDirective, { read: TemplateRef },] }],
        pageFooterTemplate: [{ type: ContentChild, args: [wizardPageFooterTemplateDirective, { read: TemplateRef },] }],
        pagedetailSelected: [{ type: Output }]
    };
    return FarrisPageDetailComponent;
}());
export { FarrisPageDetailComponent };
if (false) {
    /** @type {?} */
    FarrisPageDetailComponent.prototype.cls;
    /**
     * @type {?}
     * @private
     */
    FarrisPageDetailComponent.prototype.pageId;
    /** @type {?} */
    FarrisPageDetailComponent.prototype.pageHeaderTemplate;
    /** @type {?} */
    FarrisPageDetailComponent.prototype.pageContentTemplate;
    /** @type {?} */
    FarrisPageDetailComponent.prototype.pageFooterTemplate;
    /** @type {?} */
    FarrisPageDetailComponent.prototype.pagedetailSelected;
    /**
     * @type {?}
     * @private
     */
    FarrisPageDetailComponent.prototype.wizardSer;
    /**
     * @type {?}
     * @private
     */
    FarrisPageDetailComponent.prototype.pageSer;
    /* Skipping unhandled member: ;*/
}
