// @farris/ui-datagrid-editors/esm2015/lib/editors/datagrid-textbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-12 11:10:33
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-10-14 13:19:18
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Renderer2, ElementRef, ViewChild, Injector } from '@angular/core';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { InputDefaultOptions } from '../editor-default-options';
export class DatagridTextboxComponent extends DatagridBaseEditorDirective {
    /**
     * @param {?} render
     * @param {?} el
     * @param {?} injector
     */
    constructor(render, el, injector) {
        super(render, el, injector);
        this.inputType = 'text';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.options = Object.assign({}, InputDefaultOptions, this.options);
        this.inputElement = this.instance.nativeElement;
        if (this.options.isPassword) {
            this.inputType = 'password';
        }
        else {
            this.inputType = 'text';
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        super.ngOnDestroy();
        this.instance = null;
        this.inputElement = null;
    }
}
DatagridTextboxComponent.decorators = [
    { type: Component, args: [{
                selector: 'grid-editor-textbox',
                template: `
    <div [formGroup]="group" class="f-datagrid-cell-formgroup farris-group-auto">
        <input #input [attr.type]="inputType" class="form-control "
            [attr.placeholder]="options?.placeholder" [formControlName]="column.field"  maxlength="{{options?.maxLength}}">
    </div>
    `
            }] }
];
/** @nocollapse */
DatagridTextboxComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: Injector }
];
DatagridTextboxComponent.propDecorators = {
    instance: [{ type: ViewChild, args: ['input',] }]
};
if (false) {
    /** @type {?} */
    DatagridTextboxComponent.prototype.instance;
    /** @type {?} */
    DatagridTextboxComponent.prototype.inputType;
}
