// @farris/ui-datagrid-editors/esm2015/lib/editors/datagrid-inputgroup.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injector } from '@angular/core';
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-14 11:41:00
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-10-14 13:04:50
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Renderer2, ElementRef, ViewChild, Output } from '@angular/core';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { InputGroupComponent } from '@farris/ui-input-group';
import { InputGroupDefaultOptions } from '../editor-default-options';
export class DatagridInputGroupComponent extends DatagridBaseEditorDirective {
    /**
     * @param {?} render
     * @param {?} el
     * @param {?} injector
     */
    constructor(render, el, injector) {
        super(render, el, injector);
        this.injector = injector;
        this.clear = new EventEmitter();
        this.clickHandle = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.inputElement = this.instance.textbox.nativeElement;
        this.options = Object.assign({}, InputGroupDefaultOptions, this.options);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClear($event) {
        if (this.options.clear) {
            this.options.clear.call(this, this.eParams($event));
        }
        else {
            this.clear.emit();
        }
        this.instance.cd.detectChanges();
        return false;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClick($event) {
        if (this.options.clickHandle) {
            this.options.clickHandle.call(this, this.eParams($event));
        }
        else {
            this.clickHandle.emit();
        }
        this.instance.cd.detectChanges();
        return false;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onValueChange($event) {
        if (this.options.valueChange) {
            this.options.valueChange.call(this, this.eParams($event));
        }
    }
    /**
     * @private
     * @param {?} $event
     * @return {?}
     */
    eParams($event) {
        /** @type {?} */
        const p = this.eventParams($event);
        p['$event'] = $event;
        p['instance'] = this.instance;
        p['editor'] = this;
        p.value = $event ? $event.value : '';
        return p;
    }
}
DatagridInputGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'grid-editor-inputgroup',
                template: `
    <div [formGroup]="group" class="f-datagrid-cell-formgroup farris-group-auto">
        <input-group #ig style="width: 100%"
            (clear)="onClear($event)"
            (clickHandle)="onClick($event)"
            (valueChange)="onValueChange($event)"
            [readonly] = "options.readonly"
            [editable] = "options.editable"
            [enableClear] = "options.showClear"
            [groupText] = "options.groupText"
            [isPassword] = "options.isPassword"
            [textAlign]="options.textAlign"
            [enableViewPassword]="options.enableViewPassword"
            [formControlName]="column.field"
            [maxLength]="options.maxLength"
            ></input-group>
    </div>
    `
            }] }
];
/** @nocollapse */
DatagridInputGroupComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: Injector }
];
DatagridInputGroupComponent.propDecorators = {
    clear: [{ type: Output }],
    clickHandle: [{ type: Output }],
    instance: [{ type: ViewChild, args: ['ig',] }]
};
if (false) {
    /** @type {?} */
    DatagridInputGroupComponent.prototype.clear;
    /** @type {?} */
    DatagridInputGroupComponent.prototype.clickHandle;
    /** @type {?} */
    DatagridInputGroupComponent.prototype.instance;
    /** @type {?} */
    DatagridInputGroupComponent.prototype.injector;
}
