// @farris/ui-datagrid-editors/esm2015/lib/editors/datagrid-datepicker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Renderer2, ElementRef, NgZone, ViewChild, Injector } from '@angular/core';
import { DatagridBaseEditorDirective } from './../datagrid-base-editor.directive';
import { FarrisDatepickerComponent } from '@farris/ui-datepicker';
import { DatePickerDefaultOptions } from '../editor-default-options';
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-14 11:40:36
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-10-14 12:58:14
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
export class DatagridDatepickerComponent extends DatagridBaseEditorDirective {
    /**
     * @param {?} render
     * @param {?} el
     * @param {?} ngzone
     * @param {?} injector
     */
    constructor(render, el, ngzone, injector) {
        super(render, el, injector);
        this.ngzone = ngzone;
        this.injector = injector;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.options = Object.assign({}, DatePickerDefaultOptions, this.options);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.inputElement = this.instance.dateInput ? this.instance.dateInput.nativeElement : this.el.nativeElement.querySelector('input');
        super.ngAfterViewInit();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClear($event) {
        if (this.options.clear) {
            this.options.clear(this.eventPrams($event));
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onValueChange($event) {
        if (this.options.valueChange) {
            this.options.valueChange(this.eventPrams($event));
        }
        if (this.inputElement) {
            this.inputElement.focus();
        }
    }
    /**
     * @private
     * @param {?} $event
     * @return {?}
     */
    eventPrams($event) {
        /** @type {?} */
        const p = this.eventParams($event);
        p['instance'] = this.instance;
        p['editor'] = this;
        return p;
    }
}
DatagridDatepickerComponent.decorators = [
    { type: Component, args: [{
                selector: 'grid-editor-datepicker',
                template: `
    <div [formGroup]="group" class="f-datagrid-cell-formgroup farris-group-auto">
        <farris-datepicker
            #datepicker
            style="width: 100%"
            [formControlName]="column.field"
            [readonly]="options.readonly"
            [editable]="options.editable"
            [locale]="options.locale"
            [dateRange]="options.dateRange"
            [showTime]="options.showTime"
            [showType]="options.showType"
            [dateFormat]="options.dateFormat"
            [returnFormat]="options.returnFormat"
            [returnType]="options.returnType"
            [useDefault]="options.useDefault"
            [placeholder]="options.placeholder"
            [maxDate]="options.maxDate"
            [minDate]="options.minDate"
            [dateRangeDatesDelimiter]="options.dateRangeDatesDelimiter"
            [showWeekNumbers]="options?.showWeekNumbers"
            [shortcuts]="[]"
            (clear)="onClear($event)"
            (valueChange)="onValueChange($event)"
            [viewType]="options.viewType"
            [showPresent]="options.showPresent"
            [hourStep]="options.hourStep"
            [minuteStep]="options.minuteStep"
            [secondStep]="options.secondStep"
            [firstDayOfWeek]="options.firstDayOfWeek"
        ></farris-datepicker>
    </div>


    `
            }] }
];
/** @nocollapse */
DatagridDatepickerComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: NgZone },
    { type: Injector }
];
DatagridDatepickerComponent.propDecorators = {
    instance: [{ type: ViewChild, args: ['datepicker',] }]
};
if (false) {
    /** @type {?} */
    DatagridDatepickerComponent.prototype.instance;
    /**
     * @type {?}
     * @private
     */
    DatagridDatepickerComponent.prototype.ngzone;
    /** @type {?} */
    DatagridDatepickerComponent.prototype.injector;
}
