// @farris/ui-datagrid-editors/esm2015/lib/editors/datagrid-filtereditor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Renderer2, ElementRef, NgZone, Injector, ViewChild } from '@angular/core';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { FilterComponent } from '@farris/ui-filter-editor';
import { FilterEditorDefaultOptions } from '../editor-default-options';
export class DatagridFilterEditorComponent extends DatagridBaseEditorDirective {
    /**
     * @param {?} render
     * @param {?} el
     * @param {?} ngzone
     * @param {?} injector
     */
    constructor(render, el, ngzone, injector) {
        super(render, el, injector);
        this.ngzone = ngzone;
        this.injector = injector;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.options = Object.assign({}, FilterEditorDefaultOptions, this.options);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.inputElement = this.instance.textbox.nativeElement;
        super.ngAfterViewInit();
    }
}
DatagridFilterEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'datagrid-filter-editor',
                template: `
    <div [formGroup]="group" class="f-datagrid-cell-formgroup farris-group-auto">
        <filter #filterEditor
            style="width: 100%"
            [formControlName]="column.field"
            [columns]="options.columns"
            [showCode]="options.showCode"
            [showSql]="options.showSql"
            [disabled]="options.disabled"
            [readonly]="options.readonly"
            [editable]="options.editable"
            [enableClear]="options.enableClear"
            [fieldEditor]="options.fieldEditor"
            [enableExpress]="options.enableExpress"
            [returnType]="options.returnType"
            (showExpress)="options?.onshowExpress($event)"></filter>
    </div>
    `
            }] }
];
/** @nocollapse */
DatagridFilterEditorComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: NgZone },
    { type: Injector }
];
DatagridFilterEditorComponent.propDecorators = {
    instance: [{ type: ViewChild, args: ['filterEditor',] }]
};
if (false) {
    /** @type {?} */
    DatagridFilterEditorComponent.prototype.instance;
    /**
     * @type {?}
     * @private
     */
    DatagridFilterEditorComponent.prototype.ngzone;
    /** @type {?} */
    DatagridFilterEditorComponent.prototype.injector;
}
