// @farris/ui-datagrid-editors/esm2015/lib/editors/datagrid-enumeditor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Renderer2, ElementRef, NgZone, Injector, ViewChild } from '@angular/core';
import { EnumEditorInputComponent } from '@farris/ui-enum-editor';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { EnumEditorDefaultOptions } from '../editor-default-options';
export class DatagridEnumEditorComponent extends DatagridBaseEditorDirective {
    /**
     * @param {?} render
     * @param {?} el
     * @param {?} ngzone
     * @param {?} injector
     */
    constructor(render, el, ngzone, injector) {
        super(render, el, injector);
        this.ngzone = ngzone;
        this.injector = injector;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.options = Object.assign({}, EnumEditorDefaultOptions, this.options);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.inputElement = this.instance.inputGroupRef.textbox.nativeElement;
        super.ngAfterViewInit();
    }
}
DatagridEnumEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'datagrid-enum-editor',
                template: `
    <div [formGroup]="group" class="f-datagrid-cell-formgroup farris-group-auto">
    <enum-editor-input #eei
        style="width: 100%"
        [formControlName]="column.field"
        [readonly]="options.readonly"
        [editable]="options.editable"
        [disabled]="options.disabled"
        [enableClear]="options.enableClear"
        [outType]="options.outType"
        [textField]="options.textField"
        [valueField]="options.valueField">
    </enum-editor-input>
    </div>
    `
            }] }
];
/** @nocollapse */
DatagridEnumEditorComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: NgZone },
    { type: Injector }
];
DatagridEnumEditorComponent.propDecorators = {
    instance: [{ type: ViewChild, args: ['eei',] }]
};
if (false) {
    /** @type {?} */
    DatagridEnumEditorComponent.prototype.instance;
    /**
     * @type {?}
     * @private
     */
    DatagridEnumEditorComponent.prototype.ngzone;
    /** @type {?} */
    DatagridEnumEditorComponent.prototype.injector;
}
