// @farris/ui-datagrid-editors/esm2015/lib/editors/datagrid-timer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-14 11:41:00
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-08-24 15:33:24
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Renderer2, ElementRef, Injector, ViewChild } from '@angular/core';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { TimePickerComponent } from '@farris/ui-time-picker';
import { TimePickerDefaultOptions } from '../editor-default-options';
export class DatagridTimerComponent extends DatagridBaseEditorDirective {
    /**
     * @param {?} render
     * @param {?} el
     * @param {?} injector
     */
    constructor(render, el, injector) {
        super(render, el, injector);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.inputElement = this.instance.timeInput.nativeElement;
        this.options = Object.assign({}, TimePickerDefaultOptions, this.options);
        if (this.options && !this.options.valueChange) {
            this.options.valueChange = (/**
             * @param {?} val
             * @return {?}
             */
            (val) => { });
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onValueChange($event) {
        /** @type {?} */
        const p = this.eventParams($event);
        p.value = this.instance.value;
        p['instance'] = this.instance;
        p['editor'] = this;
        this.options.valueChange(p);
    }
}
DatagridTimerComponent.decorators = [
    { type: Component, args: [{
                selector: 'grid-editor-timer',
                template: `
    <div [formGroup]="group" class="f-datagrid-cell-formgroup farris-group-auto" >
        <farris-time-picker #timepicker
            style="width: 100%"
            [formControlName]="column.field"
            [readonly]= "options.readonly"
            [editable]= "options.editable"
            [hourStep] = "options.hourStep"
            [minuteStep] = "options.minuteStep"
            [secondStep] = "options.secondStep"
            [format]="options.format"
            [placeholder]="options.placeholder"
            [use12Hours]="options.use12Hours"
            (valueChange)="onValueChange($event)"
            [showHeader]="options.showHeader">
        </farris-time-picker>
    </div>
    `
            }] }
];
/** @nocollapse */
DatagridTimerComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: Injector }
];
DatagridTimerComponent.propDecorators = {
    instance: [{ type: ViewChild, args: ['timepicker',] }]
};
if (false) {
    /** @type {?} */
    DatagridTimerComponent.prototype.instance;
}
