// @farris/ui-datagrid-editors/esm2015/lib/editors/datagrid-numberspinner.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-14 11:41:00
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-10-22 14:25:39
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Renderer2, ElementRef, Injector, ViewChild } from '@angular/core';
import { NumberSpinnerComponent } from '@farris/ui-number-spinner';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { NumberSpinnerDefaultOptions } from '../editor-default-options';
export class DatagridNumberSpinnerComponent extends DatagridBaseEditorDirective {
    /**
     * @param {?} render
     * @param {?} el
     * @param {?} injector
     */
    constructor(render, el, injector) {
        super(render, el, injector);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.options = Object.assign({}, NumberSpinnerDefaultOptions, this.options);
        if (!this.options.textAlign) {
            this.options.textAlign = this.column.align || 'left';
        }
        this.inputElement = this.instance.input.nativeElement;
        // this.stopPropagation = false;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onValueChange($event) {
        if (this.options.valueChange) {
            this.options.valueChange(this.eventPrams($event));
        }
    }
    /**
     * @private
     * @param {?} $event
     * @return {?}
     */
    eventPrams($event) {
        /** @type {?} */
        const p = this.eventParams($event);
        p['instance'] = this.instance;
        p['editor'] = this;
        return p;
    }
}
DatagridNumberSpinnerComponent.decorators = [
    { type: Component, args: [{
                selector: 'grid-editor-numberspinner',
                template: `
    <div [formGroup]="group" class="f-datagrid-cell-formgroup farris-group-auto">
        <datagrid-tooltip [control]="formControl" [tooltipPosition]="'top-left'" [message]="errorMessage">
            <farris-number-spinner #num [step]="options.step"
                [max]="options.max"
                [min]="options.min"
                [canNull]="options.canNull"
                [precision]="options.precision"
                [editable]="options.editable"
                [bigNumber]="options.bigNumber"
                [showButton]="options.showButton"
                [showZero]="options.showZero"
                [groupSeparator]="options.groupSeparator"
                (valueChange)="onValueChange($event)"
                [textAlign]="options.textAlign"
                [formControlName]="column.field" style="width: 100%"></farris-number-spinner>
        </datagrid-tooltip>
    </div>
    `
            }] }
];
/** @nocollapse */
DatagridNumberSpinnerComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: Injector }
];
DatagridNumberSpinnerComponent.propDecorators = {
    instance: [{ type: ViewChild, args: ['num',] }]
};
if (false) {
    /** @type {?} */
    DatagridNumberSpinnerComponent.prototype.instance;
}
