// @farris/ui-datagrid-editors/esm2015/lib/editors/datagrid-switch.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-08-14 11:41:00
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-08-24 15:33:24
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, Renderer2, ElementRef, Injector, ViewChild } from '@angular/core';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { SwitchComponent } from '@farris/ui-switch';
import { SwitchDefaultOptions } from '../editor-default-options';
export class DatagridSwitchComponent extends DatagridBaseEditorDirective {
    /**
     * @param {?} render
     * @param {?} el
     * @param {?} injector
     */
    constructor(render, el, injector) {
        super(render, el, injector);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.options = Object.assign({}, SwitchDefaultOptions, this.options);
        if (this.options && !this.options.valueChange) {
            this.options.valueChange = (/**
             * @param {?} val
             * @return {?}
             */
            (val) => { });
        }
        /** @type {?} */
        const align = this.column.align || 'left';
        if (align === 'left') {
            this.render.setStyle(this.ctrlBoxEl.nativeElement, 'justify-content', align);
            this.render.setStyle(this.ctrlBoxEl.nativeElement, 'padding-left', '0.75rem');
        }
        this.inputElement = this.instance.switchElRef.nativeElement;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onValueChange($event) {
        /** @type {?} */
        const p = this.eventParams($event);
        p['checked'] = this.instance.checked;
        p.value = this.instance['getValue'](this.instance.checked);
        p['instance'] = this.instance;
        p['editor'] = this;
        this.options.valueChange(p);
    }
}
DatagridSwitchComponent.decorators = [
    { type: Component, args: [{
                selector: 'grid-editor-switch',
                template: `
    <div [formGroup]="group" class="f-datagrid-cell-formgroup farris-group-auto" style="line-height:100%" #ctrlbox hidefocus="true">
        <farris-switch #switch [formControlName]="column.field"
        [square]="options.square"
        [readonly]="options.readonly"
        [disabled]="options.disabled"
        [checkedLabel]="options.checkedLabel"
        [uncheckedLabel]="options.uncheckedLabel"
        [trueValue]="options?.trueValue"
        [falseValue]="options?.falseValue"
        (valueChange)="onValueChange($event)"></farris-switch>
    </div>
    `
            }] }
];
/** @nocollapse */
DatagridSwitchComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: Injector }
];
DatagridSwitchComponent.propDecorators = {
    ctrlBoxEl: [{ type: ViewChild, args: ['ctrlbox',] }],
    instance: [{ type: ViewChild, args: ['switch',] }]
};
if (false) {
    /** @type {?} */
    DatagridSwitchComponent.prototype.ctrlBoxEl;
    /** @type {?} */
    DatagridSwitchComponent.prototype.instance;
}
