// @farris/ui-datagrid-editors/esm2015/lib/editors/datagrid-languagetextbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Renderer2, ElementRef, Injector, ViewChild } from '@angular/core';
import { LanguageTextboxComponent } from '@farris/ui-language-textbox';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { LanguageTextboxDefaultOptions } from '../editor-default-options';
export class DatagridLanguageTextboxComponent extends DatagridBaseEditorDirective {
    /**
     * @param {?} render
     * @param {?} el
     * @param {?} injector
     */
    constructor(render, el, injector) {
        super(render, el, injector);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.inputElement = this.instance.input.textbox.nativeElement;
        this.options = Object.assign({}, LanguageTextboxDefaultOptions, this.options);
        if (this.options && !this.options.valueChange) {
            this.options.valueChange = (/**
             * @param {?} val
             * @return {?}
             */
            (val) => { });
        }
    }
}
DatagridLanguageTextboxComponent.decorators = [
    { type: Component, args: [{
                selector: 'grid-editor-language-textbox',
                template: `
    <div [formGroup]="group" class="f-datagrid-cell-formgroup farris-group-auto" style="line-height:100%">
        <language-textbox #ins style="width: 100%"
            [languages]="options.languages" [formControlName]="column.field"
            [enableClear]="options.enableClear"
            [readonly]="options.readonly"
            [currentLanguage]="options.currentLanguage"
            ></language-textbox>
    </div>
    `
            }] }
];
/** @nocollapse */
DatagridLanguageTextboxComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: Injector }
];
DatagridLanguageTextboxComponent.propDecorators = {
    instance: [{ type: ViewChild, args: ['ins',] }]
};
if (false) {
    /** @type {?} */
    DatagridLanguageTextboxComponent.prototype.instance;
}
