// @farris/ui-datagrid-editors/esm2015/lib/editors/datagrid-sorteditor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Renderer2, ElementRef, NgZone, Injector, ViewChild } from '@angular/core';
import { SorterComponent } from '@farris/ui-sort-editor';
import { DatagridBaseEditorDirective } from '../datagrid-base-editor.directive';
import { SortEditorDefaultOptions } from '../editor-default-options';
export class DatagridSortEditorComponent extends DatagridBaseEditorDirective {
    /**
     * @param {?} render
     * @param {?} el
     * @param {?} ngzone
     * @param {?} injector
     */
    constructor(render, el, ngzone, injector) {
        super(render, el, injector);
        this.ngzone = ngzone;
        this.injector = injector;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.options = Object.assign({}, SortEditorDefaultOptions, this.options);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.inputElement = this.instance.textbox.nativeElement;
        super.ngAfterViewInit();
    }
}
DatagridSortEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'datagrid-sort-editor',
                template: `
    <div [formGroup]="group" class="f-datagrid-cell-formgroup farris-group-auto">
        <sorter #sorter
            style="width: 100%"
            [formControlName]="column.field"
            [columns]="options.columns"
            [fieldEditor]="options.fieldEditor"
            [enableClear]="options.enableClear"
            [disabled]="options.disabled"
            [readonly]="options.readonly"
            [editable]="options.editable"
            [returnType]="options.returnType"
        ></sorter>
    </div>
    `
            }] }
];
/** @nocollapse */
DatagridSortEditorComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: NgZone },
    { type: Injector }
];
DatagridSortEditorComponent.propDecorators = {
    instance: [{ type: ViewChild, args: ['sorter',] }]
};
if (false) {
    /** @type {?} */
    DatagridSortEditorComponent.prototype.instance;
    /**
     * @type {?}
     * @private
     */
    DatagridSortEditorComponent.prototype.ngzone;
    /** @type {?} */
    DatagridSortEditorComponent.prototype.injector;
}
