// @gsp-wf/wf-task-handler/esm5/lib/components/urge-handle/urge-handle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var UrgeHandleComponent = /** @class */ (function () {
    function UrgeHandleComponent() {
        this.placeholder = '';
        this.showMessageSubject = false; // 是否显示消息主题
        // 消息内容文本
        this.subjectEditable = false; // 消息主题是否可编辑
        // 消息主题是否可编辑
        this.contentEditable = false; // 消息内容是否可编辑
    }
    /**
     * @return {?}
     */
    UrgeHandleComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    UrgeHandleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'urge-handle',
                    template: "<div class=\"wf-task-handler-help-div\">\r\n  <div class=\"wf-task-handler-help-text\">\r\n    {{ \"comment.static.messageSubject\" | translate }}\r\n  </div>\r\n  <div class=\"farris-input-wrap\">\r\n    <input\r\n      class=\"form-control\"\r\n      type=\"text\"\r\n      value=\"subject\"\r\n      name=\"subject\"\r\n      [placeholder]=\"placeholder\"\r\n      [(ngModel)]=\"urgeMessageSubject\"\r\n      [readOnly]=\"!subjectEditable\"\r\n    />\r\n  </div>\r\n</div>\r\n\r\n<div class=\"wf-task-handler-help-div\">\r\n  <div class=\"wf-task-handler-help-text\">\r\n    {{ \"comment.static.messageContent\" | translate }}\r\n  </div>\r\n  <textarea\r\n    class=\"form-control content-textarea\"\r\n    [(ngModel)]=\"urgeMessageContent\"\r\n    [attr.placeholder]=\"placeholder\"\r\n    [style.height.px]=\"textAreaHeight\"\r\n    [readOnly]=\"!contentEditable\"\r\n  ></textarea>\r\n</div>\r\n",
                    styles: ["", ".wf-task-handler-help-div{font-size:13px;color:#2d2f33;margin-bottom:14px}.wf-task-handler-help-div-submit{margin-bottom:14px}.wf-task-handler-help-text{padding-bottom:8px;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.wf-task-batch-selected{padding-left:8px}.wf-task-batch-rows{font-size:13px;font-weight:500;padding:0 4px;color:#2a87ff}.wf-task-batch-time{font-size:13px;font-weight:500;padding:0 4px;color:#424347}.wf-task-batch-message{display:flex;background:#edf9ff;border:1px solid rgba(196,235,255,1);border-radius:8px;margin-bottom:6px;align-items:center;font-size:12px;padding:4px 10px}.wf-submit-section{padding:4px 24px 0}.wf-submit-item{display:flex;align-items:baseline}.wf-submit-item-label{padding-right:10px;width:88px;text-align:right}.wf-task-splitter{display:flex;align-items:center;color:rgba(0,0,0,.4);font-size:12px;margin-bottom:10px}.wf-task-splitter-line{border-bottom:1px dashed rgba(0,0,0,.1);flex-grow:1}"]
                }] }
    ];
    /** @nocollapse */
    UrgeHandleComponent.ctorParameters = function () { return []; };
    UrgeHandleComponent.propDecorators = {
        placeholder: [{ type: Input }],
        showMessageSubject: [{ type: Input }],
        textAreaHeight: [{ type: Input }],
        urgeMessageSubject: [{ type: Input }],
        urgeMessageContent: [{ type: Input }],
        subjectEditable: [{ type: Input }],
        contentEditable: [{ type: Input }]
    };
    return UrgeHandleComponent;
}());
export { UrgeHandleComponent };
if (false) {
    /** @type {?} */
    UrgeHandleComponent.prototype.placeholder;
    /** @type {?} */
    UrgeHandleComponent.prototype.showMessageSubject;
    /** @type {?} */
    UrgeHandleComponent.prototype.textAreaHeight;
    /** @type {?} */
    UrgeHandleComponent.prototype.urgeMessageSubject;
    /** @type {?} */
    UrgeHandleComponent.prototype.urgeMessageContent;
    /** @type {?} */
    UrgeHandleComponent.prototype.subjectEditable;
    /** @type {?} */
    UrgeHandleComponent.prototype.contentEditable;
}
