// @gsp-svc/filtercondition/esm5/lib/filter-condition.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { LOCALE_ID, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FilterConditionComponent } from './filter-condition/filter-condition.component';
import { DropDownsModule } from '@progress/kendo-angular-dropdowns';
import { TreeTableModule } from '@farris/ui-treetable';
import { ButtonModule } from '@progress/kendo-angular-buttons';
import { DropDownButtonModule } from '@progress/kendo-angular-buttons';
import { InputsModule } from '@progress/kendo-angular-inputs';
import { FormsModule } from '@angular/forms';
import { PopupModule } from '@progress/kendo-angular-popup';
import { FilterExpressionComponent } from './filter-expression/filter-expression.component';
import { ExpressionModule } from '@gsp-svc/expression';
import { ModalModule, BsModalService } from '@farris/ui-modal';
import { FDropdownDirectiveTypeModule } from '@farris/ui-dropdown';
import { ComboListModule } from '@farris/ui-combo-list';
import { InputGroupModule } from '@farris/ui-input-group';
import { NotifyModule } from '@farris/ui-notify';
import { FilterConditionLocalePipe } from './filter-condition-locale.pipe';
var ɵ0 = localStorage.getItem('languageCode') || 'zh-CHS';
var FilterConditionModule = /** @class */ (function () {
    function FilterConditionModule() {
    }
    FilterConditionModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        DropDownsModule,
                        ButtonModule,
                        DropDownButtonModule,
                        InputsModule,
                        FormsModule,
                        PopupModule,
                        ComboListModule,
                        ExpressionModule,
                        TreeTableModule,
                        InputGroupModule,
                        NotifyModule,
                        FDropdownDirectiveTypeModule,
                        ModalModule.forRoot(),
                    ],
                    declarations: [FilterConditionComponent, FilterExpressionComponent, FilterConditionLocalePipe],
                    entryComponents: [FilterConditionComponent, FilterExpressionComponent],
                    exports: [FilterConditionComponent, FilterExpressionComponent],
                    providers: [
                        BsModalService,
                        { provide: LOCALE_ID, useValue: ɵ0 }
                    ]
                },] }
    ];
    return FilterConditionModule;
}());
export { FilterConditionModule };
export { ɵ0 };
