// @gsp-svc/filtercondition/esm5/lib/property-entity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 属性实体
 */
var /**
 * 属性实体
 */
PropertyEntity = /** @class */ (function () {
    function PropertyEntity() {
    }
    return PropertyEntity;
}());
/**
 * 属性实体
 */
export { PropertyEntity };
if (false) {
    /** @type {?} */
    PropertyEntity.prototype.name;
    /** @type {?} */
    PropertyEntity.prototype.displayName;
    /** @type {?} */
    PropertyEntity.prototype.propertyDataType;
}
/** @enum {number} */
var PropertyDataType = {
    String: 0,
    Int16: 1,
    Int32: 2,
    Int64: 3,
    Decimal: 4,
    Double: 5,
    Byte: 6,
    Char: 7,
    Boolean: 8,
    DateTime: 9,
    Array: 10,
};
export { PropertyDataType };
PropertyDataType[PropertyDataType.String] = 'String';
PropertyDataType[PropertyDataType.Int16] = 'Int16';
PropertyDataType[PropertyDataType.Int32] = 'Int32';
PropertyDataType[PropertyDataType.Int64] = 'Int64';
PropertyDataType[PropertyDataType.Decimal] = 'Decimal';
PropertyDataType[PropertyDataType.Double] = 'Double';
PropertyDataType[PropertyDataType.Byte] = 'Byte';
PropertyDataType[PropertyDataType.Char] = 'Char';
PropertyDataType[PropertyDataType.Boolean] = 'Boolean';
PropertyDataType[PropertyDataType.DateTime] = 'DateTime';
PropertyDataType[PropertyDataType.Array] = 'Array';
