// @gsp-svc/filtercondition/esm5/lib/property-adapter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FilterInjectedType } from "./injected-entity";
import { NavigationType } from "./navigation-property-entity";
import { PropertyDataType } from ".";
import { NewDataType, NewNavigationType } from "./NewSchema";
var PropertyAdapter = /** @class */ (function () {
    function PropertyAdapter() {
    }
    /**
     * 属性适配器，注入多级实体转化为Array
     * @param injectedEntity 注入的实体
     * @param schemaMap 注入的schema
     */
    /**
     * 属性适配器，注入多级实体转化为Array
     * @param {?} injectedEntity 注入的实体
     * @param {?} schemaMap 注入的schema
     * @return {?}
     */
    PropertyAdapter.prototype.adapter = /**
     * 属性适配器，注入多级实体转化为Array
     * @param {?} injectedEntity 注入的实体
     * @param {?} schemaMap 注入的schema
     * @return {?}
     */
    function (injectedEntity, schemaMap) {
        //根据注入schema名得到schema
        /** @type {?} */
        var schema = schemaMap.get(injectedEntity.schemaName);
        //根据注入类型分别进行转化
        switch (injectedEntity.injectedType) {
            case FilterInjectedType.EntitySet:
                {
                    /** @type {?} */
                    var entitySet = schema.entitySets.find((/**
                     * @param {?} entitySet
                     * @return {?}
                     */
                    function (entitySet) { return entitySet.name == injectedEntity.name; }));
                    return this.convertEntitySet(entitySet, schema);
                }
            case FilterInjectedType.EntityType:
                {
                    /** @type {?} */
                    var entityType = schema.entityTypes.find((/**
                     * @param {?} entityType
                     * @return {?}
                     */
                    function (entityType) { return entityType.name == injectedEntity.name; }));
                    return this.convertEntityType(entityType, schema);
                }
            default: break;
        }
    };
    /**
     * 将实体类转化成Arrray
     * @param entityType 被转换的实体类
     */
    /**
     * 将实体类转化成Arrray
     * @param {?} entityType 被转换的实体类
     * @param {?} schema
     * @return {?}
     */
    PropertyAdapter.prototype.convertEntityType = /**
     * 将实体类转化成Arrray
     * @param {?} entityType 被转换的实体类
     * @param {?} schema
     * @return {?}
     */
    function (entityType, schema) {
        var _this = this;
        /** @type {?} */
        var propertyList = Array();
        //常规属性转化
        entityType.properties.forEach((/**
         * @param {?} property
         * @return {?}
         */
        function (property) {
            /** @type {?} */
            var propertyListItem = {
                text: property.displayName,
                propertyDataType: property.propertyDataType,
                value: property.name,
                parentItem: null
            };
            propertyList.push(propertyListItem);
        }));
        //导航属性转化
        entityType.navigationProperties.forEach((/**
         * @param {?} navigationProperty
         * @return {?}
         */
        function (navigationProperty) {
            //根据导航属性类型进行转化
            switch (navigationProperty.navigationType) {
                //若导航属性为EntitySet
                case NavigationType.EntitySet:
                    {
                        //根据导航属性名找到schema中的EntitySet
                        /** @type {?} */
                        var entitySet = schema.entitySets.find((/**
                         * @param {?} entitySet
                         * @return {?}
                         */
                        function (entitySet) { return entitySet.name == navigationProperty.navigationEntity; }));
                        //子实体集转化为Array
                        _this.convertEntitySet(entitySet, schema).forEach((/**
                         * @param {?} element
                         * @return {?}
                         */
                        function (element) {
                            //若子实体中的每个属性父项不为空，则在父项名前添加该导航属性名，例如OrderItem前加Order为Orders.OrderItem
                            if (element.parentItem != null) {
                                element.parentItem = navigationProperty.name + "." + element.parentItem;
                            }
                            //若父项为空，则设置该属性父项为导航属性名
                            else {
                                element.parentItem = navigationProperty.name;
                            }
                            //设置该属性的显示名称
                            element.text = navigationProperty.displayName + "." + element.text;
                            propertyList.push(element);
                        }));
                    }
                    break;
                //若导航属性类型为EntityType
                case NavigationType.EntityType:
                    {
                        //根据导航属性名找到schema中的EntityType
                        /** @type {?} */
                        var entityType_1 = schema.entityTypes.find((/**
                         * @param {?} entityType
                         * @return {?}
                         */
                        function (entityType) { return entityType.name == navigationProperty.navigationEntity; }));
                        //子实体转化为Array
                        _this.convertEntityType(entityType_1, schema).forEach((/**
                         * @param {?} element
                         * @return {?}
                         */
                        function (element) {
                            //若子实体中的每个属性父项不为空，则在父项名前添加该导航属性名，例如OrderItem前加Order为Order.OrderItem
                            if (element.parentItem != null) {
                                element.parentItem = navigationProperty.name + "." + element.parentItem;
                            }
                            //若父项为空，则设置该属性父项为导航属性名
                            else {
                                element.parentItem = navigationProperty.name;
                            }
                            //设置该属性的显示名称
                            element.text = navigationProperty.displayName + "." + element.text;
                            propertyList.push(element);
                        }));
                    }
                    break;
            }
        }));
        return propertyList;
    };
    /**
     * 实体集转化为属性列表
     * @param entitySet 待转化实体集
     * @param schema 所在schema
     */
    /**
     * 实体集转化为属性列表
     * @param {?} entitySet 待转化实体集
     * @param {?} schema 所在schema
     * @return {?}
     */
    PropertyAdapter.prototype.convertEntitySet = /**
     * 实体集转化为属性列表
     * @param {?} entitySet 待转化实体集
     * @param {?} schema 所在schema
     * @return {?}
     */
    function (entitySet, schema) {
        var _this = this;
        /** @type {?} */
        var propertyList = Array();
        /** @type {?} */
        var entityType = schema.entityTypes.find((/**
         * @param {?} entityType
         * @return {?}
         */
        function (entityType) { return entityType.name == entitySet.entityTypeName; }));
        entityType.properties.forEach((/**
         * @param {?} property
         * @return {?}
         */
        function (property) {
            /** @type {?} */
            var propertyListItem = {
                text: property.displayName,
                propertyDataType: property.propertyDataType,
                value: property.name,
                parentItem: null
            };
            propertyList.push(propertyListItem);
        }));
        if (entityType.navigationProperties.length != 0) {
            //导航属性转化
            entityType.navigationProperties.forEach((/**
             * @param {?} navigationProperty
             * @return {?}
             */
            function (navigationProperty) {
                switch (navigationProperty.navigationType) {
                    case NavigationType.EntitySet:
                        {
                            /** @type {?} */
                            var entitySet_1 = schema.entitySets.find((/**
                             * @param {?} entitySet
                             * @return {?}
                             */
                            function (entitySet) { return entitySet.name == navigationProperty.navigationEntity; }));
                            _this.convertEntitySet(entitySet_1, schema).forEach((/**
                             * @param {?} element
                             * @return {?}
                             */
                            function (element) {
                                if (element.parentItem != null) {
                                    element.parentItem = navigationProperty.name + "." + element.parentItem;
                                }
                                else {
                                    element.parentItem = navigationProperty.name;
                                }
                                element.text = navigationProperty.displayName + "." + element.text;
                                propertyList.push(element);
                            }));
                        }
                        break;
                    case NavigationType.EntityType:
                        {
                            /** @type {?} */
                            var entityType_2 = schema.entityTypes.find((/**
                             * @param {?} entityType
                             * @return {?}
                             */
                            function (entityType) { return entityType.name == navigationProperty.navigationEntity; }));
                            _this.convertEntityType(entityType_2, schema).forEach((/**
                             * @param {?} element
                             * @return {?}
                             */
                            function (element) {
                                if (element.parentItem != null) {
                                    element.parentItem = navigationProperty.name + "." + element.parentItem;
                                }
                                else {
                                    element.parentItem = navigationProperty.name;
                                }
                                element.text = navigationProperty.displayName + "." + element.text;
                                propertyList.push(element);
                            }));
                        }
                        break;
                }
            }));
        }
        return propertyList;
    };
    //==========================分割线（NewSchema）============================================
    /**
     *
     * @param injectedEntity
     * @param newSchemaMap
     */
    //==========================分割线（NewSchema）============================================
    /**
     *
     * @param {?} injectedEntity
     * @param {?} newSchemaMap
     * @return {?}
     */
    PropertyAdapter.prototype.adapterOfNewSchema = 
    //==========================分割线（NewSchema）============================================
    /**
     *
     * @param {?} injectedEntity
     * @param {?} newSchemaMap
     * @return {?}
     */
    function (injectedEntity, newSchemaMap) {
        //injectedEntity.schemaName找到Schema
        /** @type {?} */
        var schema = newSchemaMap.get(injectedEntity.schemaName);
        //injectedEntity.injectedType确定找sets还是types
        switch (injectedEntity.injectedType) {
            case FilterInjectedType.EntitySet: {
                /** @type {?} */
                var set = schema.entitySets.find((/**
                 * @param {?} set
                 * @return {?}
                 */
                function (set) { return set.code == injectedEntity.name; }));
                return this.convertNewEntitySet(set, schema);
            }
            case FilterInjectedType.EntityType: {
                /** @type {?} */
                var type = schema.entityTypes.find((/**
                 * @param {?} type
                 * @return {?}
                 */
                function (type) { return type.code == injectedEntity.name; }));
                return this.convertNewEntityType(type, schema);
            }
            default:
                break;
        }
    };
    /**
     * @description 转换EntityType
     * @param type NewEntityType
     * @param schema NewSchema
     * @returns
     */
    /**
     * \@description 转换EntityType
     * @param {?} type NewEntityType
     * @param {?} schema NewSchema
     * @return {?}
     */
    PropertyAdapter.prototype.convertNewEntityType = /**
     * \@description 转换EntityType
     * @param {?} type NewEntityType
     * @param {?} schema NewSchema
     * @return {?}
     */
    function (type, schema) {
        var _this = this;
        /** @type {?} */
        var propertyList = Array();
        //property转换
        type.properties.forEach((/**
         * @param {?} pro
         * @return {?}
         */
        function (pro) {
            /** @type {?} */
            var propertyListItem = {
                value: pro.code,
                text: pro.name,
                propertyDataType: _this.convertNewDataType2PropertyDataType(pro.dataType),
                parentItem: null
            };
            propertyList.push(propertyListItem);
        }));
        //navigationProperty转换
        type.navigationProperties.forEach((/**
         * @param {?} naPro
         * @return {?}
         */
        function (naPro) {
            switch (naPro.navigationType) {
                case NewNavigationType.EntitySet:
                    {
                        /** @type {?} */
                        var naProSet = schema.entitySets.find((/**
                         * @param {?} naProSet
                         * @return {?}
                         */
                        function (naProSet) { return naProSet.code == naPro.navigationName; }));
                        _this.convertNewEntitySet(naProSet, schema).forEach((/**
                         * @param {?} ele
                         * @return {?}
                         */
                        function (ele) {
                            if (ele.parentItem != null) {
                                ele.parentItem = naPro.code + "." + ele.parentItem;
                            }
                            else {
                                ele.parentItem = naPro.code;
                            }
                            ele.text = naPro.name + "." + ele.text;
                            propertyList.push(ele);
                        }));
                    }
                    break;
                case NewNavigationType.EntityType:
                    {
                        /** @type {?} */
                        var naProType = schema.entityTypes.find((/**
                         * @param {?} naProType
                         * @return {?}
                         */
                        function (naProType) { return naProType.code == naPro.navigationName; }));
                        _this.convertNewEntityType(naProType, schema).forEach((/**
                         * @param {?} ele
                         * @return {?}
                         */
                        function (ele) {
                            if (ele.parentItem != null) {
                                ele.parentItem = naPro.code + "." + ele.parentItem;
                            }
                            else {
                                ele.parentItem = naPro.code;
                            }
                            ele.text = naPro.name + "." + ele.text;
                            propertyList.push(ele);
                        }));
                    }
                    break;
            }
        }));
        return propertyList;
    };
    /**
     * @param {?} set
     * @param {?} schema
     * @return {?}
     */
    PropertyAdapter.prototype.convertNewEntitySet = /**
     * @param {?} set
     * @param {?} schema
     * @return {?}
     */
    function (set, schema) {
        var _this = this;
        /** @type {?} */
        var propertyList = Array();
        /** @type {?} */
        var type = schema.entityTypes.find((/**
         * @param {?} type
         * @return {?}
         */
        function (type) { return type.code == set.entityType; }));
        type.properties.forEach((/**
         * @param {?} pro
         * @return {?}
         */
        function (pro) {
            /** @type {?} */
            var propertyListItem = {
                value: pro.code,
                text: pro.name,
                propertyDataType: _this.convertNewDataType2PropertyDataType(pro.dataType),
                parentItem: null
            };
            propertyList.push(propertyListItem);
        }));
        if (type.navigationProperties.length != 0) {
            type.navigationProperties.forEach((/**
             * @param {?} naPro
             * @return {?}
             */
            function (naPro) {
                switch (naPro.navigationType) {
                    case NewNavigationType.EntitySet:
                        {
                            /** @type {?} */
                            var naProSet = schema.entitySets.find((/**
                             * @param {?} naProSet
                             * @return {?}
                             */
                            function (naProSet) { return naProSet.code == naPro.navigationName; }));
                            _this.convertNewEntitySet(naProSet, schema).forEach((/**
                             * @param {?} ele
                             * @return {?}
                             */
                            function (ele) {
                                if (ele.parentItem != null) {
                                    ele.parentItem = naPro.code + "." + ele.parentItem;
                                }
                                else {
                                    ele.parentItem = naPro.code;
                                }
                                ele.text = naPro.name + "." + ele.text;
                                propertyList.push(ele);
                            }));
                        }
                        break;
                    case NewNavigationType.EntityType:
                        {
                            /** @type {?} */
                            var naProType = schema.entityTypes.find((/**
                             * @param {?} naProType
                             * @return {?}
                             */
                            function (naProType) { return naProType.code == naPro.navigationName; }));
                            _this.convertNewEntityType(naProType, schema).forEach((/**
                             * @param {?} ele
                             * @return {?}
                             */
                            function (ele) {
                                if (ele.parentItem != null) {
                                    ele.parentItem = naPro.code + "." + ele.parentItem;
                                }
                                else {
                                    ele.parentItem = naPro.code;
                                }
                                ele.text = naPro.name + "." + ele.text;
                                propertyList.push(ele);
                            }));
                        }
                        break;
                }
            }));
        }
        return propertyList;
    };
    /**
     *
     * @param newType 统一Schema的属性
     * @returns
     */
    /**
     *
     * @param {?} newType 统一Schema的属性
     * @return {?}
     */
    PropertyAdapter.prototype.convertNewDataType2PropertyDataType = /**
     *
     * @param {?} newType 统一Schema的属性
     * @return {?}
     */
    function (newType) {
        /** @type {?} */
        var type = PropertyDataType.String;
        switch (newType) {
            case NewDataType.Boolean:
                type = PropertyDataType.Boolean;
                break;
            case NewDataType.Byte:
                type = PropertyDataType.Byte;
                break;
            case NewDataType.Date:
                type = PropertyDataType.DateTime;
                break;
            case NewDataType.DateTime:
                type = PropertyDataType.DateTime;
                break;
            case NewDataType.Decimal:
                type = PropertyDataType.Decimal;
                break;
            case NewDataType.Integer:
                type = PropertyDataType.Int32;
                break;
            case NewDataType.String:
                type = PropertyDataType.String;
                break;
        }
        return type;
    };
    return PropertyAdapter;
}());
export { PropertyAdapter };
if (false) {
    /** @type {?} */
    PropertyAdapter.prototype.entityName;
}
