// @gsp-svc/filtercondition/esm5/lib/schema.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * schema实体类
 * \@property entitySets  实体集集合
 * \@property entityTypes 实体类型集合
 */
var /**
 * schema实体类
 * \@property entitySets  实体集集合
 * \@property entityTypes 实体类型集合
 */
FilterEntitySchema = /** @class */ (function () {
    function FilterEntitySchema() {
        this.entitySets = [];
        this.entityTypes = [];
    }
    return FilterEntitySchema;
}());
/**
 * schema实体类
 * \@property entitySets  实体集集合
 * \@property entityTypes 实体类型集合
 */
export { FilterEntitySchema };
if (false) {
    /** @type {?} */
    FilterEntitySchema.prototype.entitySets;
    /** @type {?} */
    FilterEntitySchema.prototype.entityTypes;
}
/**
 * 过滤实体类型
 * 用于多级嵌套实体显示
 * \@property name 实体名(用于实体间导航使用)
 * \@property displayName 展示名称
 * \@property properties 属性集
 * \@property navigationProperties 导航属性集
 */
var /**
 * 过滤实体类型
 * 用于多级嵌套实体显示
 * \@property name 实体名(用于实体间导航使用)
 * \@property displayName 展示名称
 * \@property properties 属性集
 * \@property navigationProperties 导航属性集
 */
FilterEntityType = /** @class */ (function () {
    function FilterEntityType() {
        this.properties = [];
        this.navigationProperties = [];
    }
    return FilterEntityType;
}());
/**
 * 过滤实体类型
 * 用于多级嵌套实体显示
 * \@property name 实体名(用于实体间导航使用)
 * \@property displayName 展示名称
 * \@property properties 属性集
 * \@property navigationProperties 导航属性集
 */
export { FilterEntityType };
if (false) {
    /** @type {?} */
    FilterEntityType.prototype.name;
    /** @type {?} */
    FilterEntityType.prototype.displayName;
    /** @type {?} */
    FilterEntityType.prototype.properties;
    /** @type {?} */
    FilterEntityType.prototype.navigationProperties;
}
/**
 * 过滤实体类型集
 * 用于多级嵌套实体显示
 * \@property name 实体集名
 * \@property entityTypeName 包含的实体类型名
 */
var /**
 * 过滤实体类型集
 * 用于多级嵌套实体显示
 * \@property name 实体集名
 * \@property entityTypeName 包含的实体类型名
 */
FilterEntitySet = /** @class */ (function () {
    function FilterEntitySet() {
    }
    return FilterEntitySet;
}());
/**
 * 过滤实体类型集
 * 用于多级嵌套实体显示
 * \@property name 实体集名
 * \@property entityTypeName 包含的实体类型名
 */
export { FilterEntitySet };
if (false) {
    /** @type {?} */
    FilterEntitySet.prototype.name;
    /** @type {?} */
    FilterEntitySet.prototype.displayName;
    /** @type {?} */
    FilterEntitySet.prototype.entityTypeName;
}
