// @gsp-sys/sysmgr-common/esm2015/lib/service/sys-manager-util.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { SessionService } from '@ecp-caf/caf-common';
/**
 * 系统管理公共session、webapi等公用方法
 */
export class SysCommonUtilService {
    /**
     * @param {?} sessionService
     */
    constructor(sessionService) {
        this.sessionService = sessionService;
        /** @type {?} */
        const apiPath = '/api/runtime/sys/v1.0';
        this.userWebApi = apiPath + '/users';
        this.secLevelWebApi = apiPath + '/seclevels';
        this.userLimitWebApi = apiPath + '/userLimit';
        this.userPositionWebApi = apiPath + '/user-position';
        this.userPositionOrgWebApi = apiPath + '/user-position-org';
        this.sysOrgWebApi = apiPath + '/sysOrgs';
        this.positionWebApi = apiPath + '/positions';
        this.roleWebApi = apiPath + '/roles';
        this.positionRoleWebApi = apiPath + '/position-role';
        this.operationWebApi = apiPath + '/funcOperations';
        this.userGroupWebApi = apiPath + '/usergroups';
        this.posGroupWebApi = apiPath + '/posgroups';
        this.roleGroupWebApi = apiPath + '/rolegroups';
        this.authFieldDesignWebApi = apiPath + '/authfields';
        this.authFieldAssignWebApi = apiPath + '/assignresults';
        this.sysFuncOpPermissionWebapi = apiPath + '/sysFuncOpPermission';
        this.sysSecurityApi = apiPath + '/secmgrset';
    }
    /**
     * 获取当前用户的sessionId
     * @return {?}
     */
    getSessionId() {
        return this.sessionService.getUserSessionId();
    }
    /**
     * 构造Header
     * @return {?}
     */
    getHeader() {
        /** @type {?} */
        const options = {
            headers: {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                SessionId: this.getSessionId()
            }
        };
        return options;
    }
}
SysCommonUtilService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
SysCommonUtilService.ctorParameters = () => [
    { type: SessionService }
];
if (false) {
    /**
     * 用户webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.userWebApi;
    /**
     * 用户密级字典webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.secLevelWebApi;
    /**
     * 登录限制
     * @type {?}
     */
    SysCommonUtilService.prototype.userLimitWebApi;
    /**
     * 用户组织岗位关系webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.userPositionWebApi;
    /**
     * 用户通用岗位业务组织关系webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.userPositionOrgWebApi;
    /**
     * 系统组织webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.sysOrgWebApi;
    /**
     * 岗位webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.positionWebApi;
    /**
     * 功能组webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.roleWebApi;
    /**
     * 功能组和岗位关系webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.positionRoleWebApi;
    /**
     * 功能操作webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.operationWebApi;
    /**
     * 用户分组webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.userGroupWebApi;
    /**
     * 岗位分组webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.posGroupWebApi;
    /**
     * 功能组分组webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.roleGroupWebApi;
    /**
     * 权限字段定义时webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.authFieldDesignWebApi;
    /**
     * 权限字段分配时webapi
     * @type {?}
     */
    SysCommonUtilService.prototype.authFieldAssignWebApi;
    /** @type {?} */
    SysCommonUtilService.prototype.sysFuncOpPermissionWebapi;
    /**
     * 保密测评相关
     * @type {?}
     */
    SysCommonUtilService.prototype.sysSecurityApi;
    /**
     * @type {?}
     * @private
     */
    SysCommonUtilService.prototype.sessionService;
}
