// @gsp-sys/sysmgr-common/esm5/lib/config/sys-security-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { SysCommonUtilService } from '../service/sys-manager-util.service';
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../service/sys-manager-util.service";
/**
 * 保密测评服务类
 */
var SysSecurityConfigService = /** @class */ (function () {
    function SysSecurityConfigService(http, sysUtil) {
        this.http = http;
        this.sysUtil = sysUtil;
    }
    /** 是否开启三员 */
    /**
     * 是否开启三员
     * @return {?}
     */
    SysSecurityConfigService.prototype.isOnSecurity = /**
     * 是否开启三员
     * @return {?}
     */
    function () {
        /** @type {?} */
        var url = this.sysUtil.sysSecurityApi + '/useSecurity';
        return this.http.get(url, this.sysUtil.getHeader()).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return (/** @type {?} */ (res)); })));
    };
    /**
     * 根据当前登录用户获取保密测评的功能配置范围
     * @param menuId 功能标识
     */
    /**
     * 根据当前登录用户获取保密测评的功能配置范围
     * @param {?} menuId 功能标识
     * @return {?}
     */
    SysSecurityConfigService.prototype.getCurrentUserAppConfig = /**
     * 根据当前登录用户获取保密测评的功能配置范围
     * @param {?} menuId 功能标识
     * @return {?}
     */
    function (menuId) {
        /** @type {?} */
        var url = this.sysUtil.sysSecurityApi + '/funcobj/' + menuId;
        return this.http.get(url, this.sysUtil.getHeader());
    };
    /**
     * 根据指定用户类别获取界面显示信息-,目前只传递type="alluser"使用
     * @param menuId 功能标识
     */
    /**
     * 根据指定用户类别获取界面显示信息-,目前只传递type="alluser"使用
     * @param {?} menuId 功能标识
     * @param {?} type
     * @return {?}
     */
    SysSecurityConfigService.prototype.getAppConfigByUserType = /**
     * 根据指定用户类别获取界面显示信息-,目前只传递type="alluser"使用
     * @param {?} menuId 功能标识
     * @param {?} type
     * @return {?}
     */
    function (menuId, type) {
        /** @type {?} */
        var url = this.sysUtil.sysSecurityApi + '/funcobj/' + menuId + '/' + type;
        return this.http.get(url, this.sysUtil.getHeader());
    };
    SysSecurityConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    SysSecurityConfigService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: SysCommonUtilService }
    ]; };
    /** @nocollapse */ SysSecurityConfigService.ngInjectableDef = i0.defineInjectable({ factory: function SysSecurityConfigService_Factory() { return new SysSecurityConfigService(i0.inject(i1.HttpClient), i0.inject(i2.SysCommonUtilService)); }, token: SysSecurityConfigService, providedIn: "root" });
    return SysSecurityConfigService;
}());
export { SysSecurityConfigService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    SysSecurityConfigService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    SysSecurityConfigService.prototype.sysUtil;
}
