// @qdp/ide-cmp/esm2015/lib/service/print-integration/print-integration.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { PrintIntegrationManagerService } from '@qdp/common';
export class PrintIntegrationService {
    /**
     * @param {?} printIntegrationManager
     */
    constructor(printIntegrationManager) {
        this.printIntegrationManager = printIntegrationManager;
        this.printIntegrationInfo = new Subject();
        this.printIntegrationManager.printIntegrationInfoChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            this.printIntegrationInfo.next(value);
        }));
    }
    /**
     * @param {?} schemaId
     * @param {?} queryId
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    getPrintIntegration(schemaId, queryId, queryRelativeUrl) {
        return this.printIntegrationManager.getPrintIntegration(schemaId, queryId, queryRelativeUrl);
    }
    /**
     * @param {?} printIntegrationInfo
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    savePrintIntegration(printIntegrationInfo, queryRelativeUrl) {
        return this.printIntegrationManager.savePrintIntegration(printIntegrationInfo, queryRelativeUrl);
    }
    /**
     * @param {?} schemaId
     * @param {?} queryID
     * @param {?} queryRelativeUrl
     * @return {?}
     */
    deletPrintIntegration(schemaId, queryID, queryRelativeUrl) {
        return this.printIntegrationManager.deletePrintIntegration(schemaId, queryID, queryRelativeUrl);
    }
}
PrintIntegrationService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
PrintIntegrationService.ctorParameters = () => [
    { type: PrintIntegrationManagerService }
];
if (false) {
    /** @type {?} */
    PrintIntegrationService.prototype.printIntegrationInfo;
    /**
     * @type {?}
     * @private
     */
    PrintIntegrationService.prototype.printIntegrationManager;
}
