// @gsp-wf/wf-process-editor/esm2015/lib/views/participant/participant.info.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { AssignStrategy } from "../assign-strategy/assign-strategy";
export class ParticipantInfo {
    constructor() {
        this.assignStrategy = new AssignStrategy();
        this.humanPerformerList = new Array();
        this.potentialOwnerList = new Array();
        this.isAllScope = false;
    }
}
if (false) {
    /** @type {?} */
    ParticipantInfo.prototype.assignStrategy;
    /** @type {?} */
    ParticipantInfo.prototype.humanPerformerList;
    /** @type {?} */
    ParticipantInfo.prototype.potentialOwnerList;
    /** @type {?} */
    ParticipantInfo.prototype.isAllScope;
}
export class Participant {
}
if (false) {
    /** @type {?} */
    Participant.prototype.id;
    /** @type {?} */
    Participant.prototype.type;
    /** @type {?} */
    Participant.prototype.condition;
    /** @type {?} */
    Participant.prototype.order;
}
export class UserParticipant extends Participant {
}
if (false) {
    /** @type {?} */
    UserParticipant.prototype.performerId;
    /** @type {?} */
    UserParticipant.prototype.performerName;
    /** @type {?} */
    UserParticipant.prototype.orgId;
    /** @type {?} */
    UserParticipant.prototype.orgName;
    /** @type {?} */
    UserParticipant.prototype.orgType;
    /** @type {?} */
    UserParticipant.prototype.tag;
}
export class VariableParticipant extends UserParticipant {
}
if (false) {
    /** @type {?} */
    VariableParticipant.prototype.name;
    /** @type {?} */
    VariableParticipant.prototype.varValue;
    /** @type {?} */
    VariableParticipant.prototype.varType;
    /** @type {?} */
    VariableParticipant.prototype.relation;
}
export class PositionParticipant extends UserParticipant {
}
if (false) {
    /** @type {?} */
    PositionParticipant.prototype.positionType;
    /** @type {?} */
    PositionParticipant.prototype.isRequired;
}
export class ComponentParticipant extends Participant {
}
if (false) {
    /** @type {?} */
    ComponentParticipant.prototype.name;
    /** @type {?} */
    ComponentParticipant.prototype.metadataId;
    /** @type {?} */
    ComponentParticipant.prototype.methodCode;
    /** @type {?} */
    ComponentParticipant.prototype.methodName;
    /** @type {?} */
    ComponentParticipant.prototype.actualParameters;
}
export class ReportingLineParticipant extends UserParticipant {
}
if (false) {
    /** @type {?} */
    ReportingLineParticipant.prototype.reportingLineDefId;
    /** @type {?} */
    ReportingLineParticipant.prototype.reportingLineDefName;
    /** @type {?} */
    ReportingLineParticipant.prototype.refOrgId;
    /** @type {?} */
    ReportingLineParticipant.prototype.refOrgName;
    /** @type {?} */
    ReportingLineParticipant.prototype.refUserId;
    /** @type {?} */
    ReportingLineParticipant.prototype.refEmployeeId;
    /** @type {?} */
    ReportingLineParticipant.prototype.dimension;
    /** @type {?} */
    ReportingLineParticipant.prototype.reportingReferObject;
}
export class MatrixParticipant extends UserParticipant {
}
if (false) {
    /** @type {?} */
    MatrixParticipant.prototype.matrixDefId;
    /** @type {?} */
    MatrixParticipant.prototype.matrixDefName;
    /** @type {?} */
    MatrixParticipant.prototype.conditionField1;
    /** @type {?} */
    MatrixParticipant.prototype.conditionFieldText1;
    /** @type {?} */
    MatrixParticipant.prototype.conditionField2;
    /** @type {?} */
    MatrixParticipant.prototype.conditionFieldText2;
    /** @type {?} */
    MatrixParticipant.prototype.conditionField3;
    /** @type {?} */
    MatrixParticipant.prototype.conditionFieldText3;
    /** @type {?} */
    MatrixParticipant.prototype.dataFieldAlias;
}
export class ActivityExecutorParticipant extends UserParticipant {
}
if (false) {
    /** @type {?} */
    ActivityExecutorParticipant.prototype.activityDefId;
    /** @type {?} */
    ActivityExecutorParticipant.prototype.activityDefName;
}
export class OrganizationParticipant extends Participant {
}
if (false) {
    /** @type {?} */
    OrganizationParticipant.prototype.name;
    /** @type {?} */
    OrganizationParticipant.prototype.actionScope;
    /** @type {?} */
    OrganizationParticipant.prototype.organizationLevel;
    /** @type {?} */
    OrganizationParticipant.prototype.organizationId;
    /** @type {?} */
    OrganizationParticipant.prototype.organizationCode;
    /** @type {?} */
    OrganizationParticipant.prototype.organizationName;
    /** @type {?} */
    OrganizationParticipant.prototype.includeChildren;
    /** @type {?} */
    OrganizationParticipant.prototype.parentsUntil;
    /** @type {?} */
    OrganizationParticipant.prototype.tag;
}
/** @enum {string} */
const ActionScope = {
    /**
     * 指派
     */
    AssignUser: 'AssignUser',
    /**
     * 移交
     */
    Transfer: 'Transfer',
    /**
     * 征询
     */
    AddSignFront: 'AddSignFront',
    /**
     * 加签
     */
    AddSignBehind: 'AddSignBehind',
    /**
     * 知会
     */
    Notify: 'Notify',
};
export { ActionScope };
/** @enum {string} */
const OrganizationLevel = {
    /**
     * 部门
     */
    DEPARTMENT: 'DEPARTMENT',
    /**
     * 单位
     */
    UNIT: 'UNIT',
    /**
    * 默认值
    */
    DEFAULT: 'DEFAULT',
    VARORGANIZATION: 'VARORGANIZATION',
};
export { OrganizationLevel };
export class ActualParameter {
}
if (false) {
    /** @type {?} */
    ActualParameter.prototype.code;
    /** @type {?} */
    ActualParameter.prototype.value;
    /** @type {?} */
    ActualParameter.prototype.valueText;
    /** @type {?} */
    ActualParameter.prototype.tag;
    /** @type {?} */
    ActualParameter.prototype.fetchValueMode;
    /** @type {?} */
    ActualParameter.prototype.fetchValueSetting;
}
