// @farris/ui-messager/esm2015/lib/farris-default-focus.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, NgZone } from '@angular/core';
export class FarrisDefaultFocusDirective {
    /**
     * @param {?} el
     * @param {?} ngZone
     */
    constructor(el, ngZone) {
        this.el = el;
        this.ngZone = ngZone;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.button.defaultFocus) {
            this.ngZone.runOutsideAngular((/**
             * @return {?}
             */
            () => {
                setTimeout((/**
                 * @return {?}
                 */
                () => {
                    this.el.nativeElement.focus();
                }));
            }));
        }
    }
}
FarrisDefaultFocusDirective.decorators = [
    { type: Directive, args: [{
                selector: '[farris-default-focus]'
            },] }
];
/** @nocollapse */
FarrisDefaultFocusDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NgZone }
];
FarrisDefaultFocusDirective.propDecorators = {
    button: [{ type: Input, args: ['farris-default-focus',] }]
};
if (false) {
    /** @type {?} */
    FarrisDefaultFocusDirective.prototype.button;
    /**
     * @type {?}
     * @private
     */
    FarrisDefaultFocusDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FarrisDefaultFocusDirective.prototype.ngZone;
}
