// @farris/ui-messager/esm2015/lib/messager.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { FarrisCommonModule } from '@farris/ui-common';
// import { FarrisFormsModule } from '@farris/ui-forms';
import { ModalModule } from '@farris/ui-modal';
import { LocaleModule } from '@farris/ui-locale';
import { MessagerComponent } from './messager.component';
import { MessagerService } from './messager.service';
import { MESSAGER_DEFAULT_CONFIG, MessagerDefaultConfig } from './messager.config';
import { FarrisDefaultFocusDirective } from './farris-default-focus.directive';
export class MessagerModule {
    /**
     * @param {?=} config
     * @return {?}
     */
    static forRoot(config) {
        return {
            ngModule: MessagerModule,
            providers: [
                { provide: MESSAGER_DEFAULT_CONFIG, useValue: config || MessagerDefaultConfig },
                MessagerService
            ]
        };
    }
}
MessagerModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    FarrisCommonModule.forRoot(),
                    // FarrisFormsModule,
                    LocaleModule,
                    ModalModule.forRoot()
                ],
                declarations: [
                    FarrisDefaultFocusDirective,
                    MessagerComponent
                ],
                entryComponents: [
                    MessagerComponent
                ]
            },] }
];
