// @farris/ui-messager/esm2015/lib/messager.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/**
 * @record
 */
export function MessagerConfig() { }
if (false) {
    /** @type {?|undefined} */
    MessagerConfig.prototype.width;
    /** @type {?|undefined} */
    MessagerConfig.prototype.height;
    /** @type {?|undefined} */
    MessagerConfig.prototype.title;
    /** @type {?|undefined} */
    MessagerConfig.prototype.showHeader;
}
/** @type {?} */
export const MessagerDefaultConfig = {
    // width: 340,
    // height: 220,
    title: '系统提示'
};
/** @type {?} */
export const MESSAGER_DEFAULT_CONFIG = new InjectionToken('Messager default options.');
/** @enum {number} */
const ExceptionLevel = {
    /**
     * 提示信息
     */
    Info: 0,
    /**
     * 警告
     */
    Warning: 1,
    /**
     * 错误
     */
    Error: 2,
    /**
     * 致命错误
     */
    Fatal: 3,
};
export { ExceptionLevel };
/**
 * @record
 */
export function ServerExceptionEntity() { }
if (false) {
    /** @type {?} */
    ServerExceptionEntity.prototype.Code;
    /** @type {?} */
    ServerExceptionEntity.prototype.Level;
    /** @type {?} */
    ServerExceptionEntity.prototype.Message;
    /** @type {?} */
    ServerExceptionEntity.prototype.Detail;
    /** @type {?} */
    ServerExceptionEntity.prototype.RequestId;
    /** @type {?|undefined} */
    ServerExceptionEntity.prototype.date;
    /** @type {?|undefined} */
    ServerExceptionEntity.prototype.innerMessage;
}
/**
 * @record
 */
export function HttpRequestFaildInfo() { }
if (false) {
    /** @type {?} */
    HttpRequestFaildInfo.prototype.headers;
    /** @type {?} */
    HttpRequestFaildInfo.prototype.status;
    /** @type {?} */
    HttpRequestFaildInfo.prototype.statusText;
    /** @type {?} */
    HttpRequestFaildInfo.prototype.url;
    /** @type {?} */
    HttpRequestFaildInfo.prototype.ok;
    /** @type {?} */
    HttpRequestFaildInfo.prototype.name;
    /** @type {?} */
    HttpRequestFaildInfo.prototype.message;
    /** @type {?} */
    HttpRequestFaildInfo.prototype.error;
}
