// @farris/ui-list-view/esm5/lib/template.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, TemplateRef } from '@angular/core';
var ListViewHeaderTemplateDirective = /** @class */ (function () {
    function ListViewHeaderTemplateDirective(template) {
        this.template = template;
    }
    ListViewHeaderTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[listViewHeader]'
                },] }
    ];
    /** @nocollapse */
    ListViewHeaderTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return ListViewHeaderTemplateDirective;
}());
export { ListViewHeaderTemplateDirective };
if (false) {
    /** @type {?} */
    ListViewHeaderTemplateDirective.prototype.template;
}
var ListViewTemplateDirective = /** @class */ (function () {
    function ListViewTemplateDirective(template) {
        this.template = template;
    }
    ListViewTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[listTemplate]'
                },] }
    ];
    /** @nocollapse */
    ListViewTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return ListViewTemplateDirective;
}());
export { ListViewTemplateDirective };
if (false) {
    /** @type {?} */
    ListViewTemplateDirective.prototype.template;
}
var ListViewFooterTemplateDirective = /** @class */ (function () {
    function ListViewFooterTemplateDirective(template) {
        this.template = template;
    }
    ListViewFooterTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[listViewFooter]'
                },] }
    ];
    /** @nocollapse */
    ListViewFooterTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return ListViewFooterTemplateDirective;
}());
export { ListViewFooterTemplateDirective };
if (false) {
    /** @type {?} */
    ListViewFooterTemplateDirective.prototype.template;
}
var ListViewEmptyTemplateDirective = /** @class */ (function () {
    function ListViewEmptyTemplateDirective(template) {
        this.template = template;
    }
    ListViewEmptyTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[listViewEmpty]'
                },] }
    ];
    /** @nocollapse */
    ListViewEmptyTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return ListViewEmptyTemplateDirective;
}());
export { ListViewEmptyTemplateDirective };
if (false) {
    /** @type {?} */
    ListViewEmptyTemplateDirective.prototype.template;
}
