// @farris/header-group-editor/esm2015/lib/column-group-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BsModalService } from '@farris/ui-modal';
import { HeaderGroupEditorComponent } from './header-group-editor.component';
import { Component, Input, ViewChild, EventEmitter, Output, TemplateRef, ElementRef, Renderer2, Injector, HostBinding } from '@angular/core';
import { MessagerService } from '@farris/ui-messager';
import { flatten } from 'lodash-es';
export class ColumnGroupEditorComponent {
    /**
     * @param {?} render
     * @param {?} messagerSer
     * @param {?} injector
     * @param {?} modalSer
     */
    constructor(render, messagerSer, injector, modalSer) {
        this.render = render;
        this.messagerSer = messagerSer;
        this.injector = injector;
        this.modalSer = modalSer;
        this.editorParams = {};
        this.cls = 'f-utils-absolute-all';
        this.closeModal = new EventEmitter();
        // 模态框确定后关闭，并传递参数
        this.submitModal = new EventEmitter();
        this.previewGroupColumns = [];
        this.columnsCode = [];
        this.btnTopActive = true;
        this.btnBottomActive = true;
        this.isOpen = false;
        this.fieldInfo = {};
        this.modalConfig = {
            title: '多表头编辑器',
            width: 700,
            height: 500,
            showButtons: true
        };
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (!this.fields || !this.fields.length) {
            if (this.editorParams && this.editorParams.fields && this.editorParams.fields.length) {
                this.fields = this.editorParams.fields;
            }
            else {
                console.warn('Pleasy Set Datagrid\'s fields Give it 2 me.');
            }
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onTreenodeSelect($event) {
        this.fieldInfo = $event.data;
    }
    /**
     * @private
     * @return {?}
     */
    setPreviewColumns() {
        /** @type {?} */
        const treeNodes = this.hge.getAllNodes();
        this.previewGroupColumns = this.hge.convert2Array(treeNodes);
    }
    /**
     * @return {?}
     */
    showPreview() {
        this.setPreviewColumns();
        if (!this.isOpen) {
            this.isOpen = true;
        }
    }
    /**
     * @param {?=} event
     * @return {?}
     */
    closePreview(event) {
        this.isOpen = false;
        if (event) {
            event.stopPropagation();
        }
    }
    /**
     * @return {?}
     */
    resetGroupColumns() {
        this.messagerSer.question('表头分组重置后需要重新设置。确定要重置为初始状态吗？', (/**
         * @return {?}
         */
        () => {
            this.hge.initTreeData();
        }));
    }
    /**
     * @return {?}
     */
    cancle() {
        this.closeModal.emit();
    }
    /**
     * @return {?}
     */
    ok() {
        this.setPreviewColumns();
        /** @type {?} */
        const flatgroupcols = flatten(this.previewGroupColumns);
        /** @type {?} */
        const value = this.hge.tt.serializedValue.map((/**
         * @param {?} n
         * @return {?}
         */
        n => {
            /** @type {?} */
            const parent = n.parent ? n.parent.data.id : null;
            if (n.node.data.groupHeader || n.node.data.houxunquNode) {
                /** @type {?} */
                const _col = flatgroupcols.find((/**
                 * @param {?} c
                 * @return {?}
                 */
                c => c.id === n.id));
                /** @type {?} */
                const item = Object.assign({}, n.node.data, { parent, level: n.level });
                if (_col) {
                    return Object.assign(item, { rowspan: _col ? _col.rowspan : 1,
                        colspan: _col ? _col.colspan : 1 });
                }
                return item;
            }
            else {
                return {
                    fieldRef: n.node.data.id,
                    parent,
                    level: n.level
                };
            }
        }));
        this.submitModal.emit({ value });
    }
    /**
     * @return {?}
     */
    showCode() {
        /** @type {?} */
        const treeNodes = this.hge.getAllNodes();
        this.columnsCode = this.hge.convert2Array(treeNodes);
        this.modalSer.show(this.codeTemplate, {
            title: '表头分组代码',
            width: 500, height: 400,
            resizable: false,
            showMaxButton: false,
            showButtons: false
        });
    }
}
ColumnGroupEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'column-group-editor',
                template: "\r\n<div class=\"d-flex flex-column\" style=\"height: 100%\">\r\n        <div region=\"north\" class=\"north\" style=\"height: 40px;min-height: 40px;border-bottom: 1px solid #c8c8c8; line-height: 36px;\">\r\n            <button class=\"btn btn-primary mr-1 ml-2\" (click)=\"resetGroupColumns()\">\r\n                <span class=\"f-icon f-icon-reset-sm mr-1\"></span>\u91CD\u7F6E\r\n            </button>\r\n            <button class=\"btn btn-primary mr-1\" (click)=\"showPreview()\">\r\n                <span class=\"f-icon f-icon-preview  mr-1\"></span>\u6548\u679C\u9884\u89C8\r\n            </button>\r\n            <button class=\"btn btn-primary\" (click)=\"showCode()\">\r\n                <span class=\"f-icon f-icon-source-code mr-1\"></span>\u83B7\u53D6\u4EE3\u7801\r\n            </button>\r\n        </div>\r\n        <div region=\"center\" class=\"center flex-fill\" style=\"overflow:auto;height: 100%;\">\r\n            <div class=\"d-flex flex-row flex-fill\" style=\"height: 100%\" >\r\n                <div region=\"center\" class=\"center flex-fill\" style=\"overflow:auto;\">\r\n                    <header-group-editor #hge [fields]=\"fields\" [treeData]=\"value\" (selectedNode)=\"onTreenodeSelect($event)\"></header-group-editor>\r\n                </div>\r\n                <div region=\"east\" class=\"east property-column\">\r\n\r\n                    <div style=\"padding: 10px\">\r\n                            <div class=\"farris-group-wrap\">\r\n                                <div class=\"form-group farris-form-group\">\r\n                                    <label for=\"hpinput01\" class=\"col-form-label\">\r\n                                        <span class=\"farris-label-text\">\u8868\u5934\u540D\u79F0</span>\r\n                                    </label>\r\n                                    <div class=\"farris-input-wrap\">\r\n                                        <input type=\"input\" [disabled]=\"!fieldInfo.groupHeader\" class=\"form-control\" [(ngModel)]=\"fieldInfo.caption\">\r\n                                    </div>\r\n                                </div>\r\n                            </div>\r\n                            <div class=\"farris-group-wrap\" *ngIf=\"fieldInfo.groupHeader\">\r\n                                <div class=\"form-group farris-form-group\">\r\n                                    <label for=\"hpinput01\" class=\"col-form-label\">\r\n                                        <span class=\"farris-label-text\">\u8868\u5934\u5BF9\u9F50\u65B9\u5F0F</span>\r\n                                    </label>\r\n                                    <div class=\"farris-input-wrap\">\r\n                                        <select class=\"form-control\" [(ngModel)]=\"fieldInfo.halign\">\r\n                                            <option value=\"left\">\u5DE6\u5BF9\u9F50</option>\r\n                                            <option value=\"center\">\u5C45\u4E2D\u5BF9\u9F50</option>\r\n                                            <option value=\"right\">\u53F3\u5BF9\u9F50</option>\r\n                                        </select>\r\n                                        <!-- <select class=\"form-control\" *ngIf=\"!fieldInfo.groupHeader\" [(ngModel)]=\"fieldInfo.hAlign\">\r\n                                            <option value=\"left\">\u5DE6\u5BF9\u9F50</option>\r\n                                            <option value=\"center\">\u5C45\u4E2D\u5BF9\u9F50</option>\r\n                                            <option value=\"right\">\u53F3\u5BF9\u9F50</option>\r\n                                        </select> -->\r\n                                    </div>\r\n                                </div>\r\n                            </div>\r\n                     \r\n                             <!-- <div class=\"farris-group-wrap\">\r\n                                <div class=\"form-group farris-form-group\">\r\n                                    <label for=\"hpinput01\" class=\"col-form-label\">\r\n                                        <span class=\"farris-label-text\">\u6570\u636E\u5BF9\u9F50\u65B9\u5F0F</span>\r\n                                    </label>\r\n                                    <div class=\"farris-input-wrap\">\r\n                                        <select class=\"form-control\" [(ngModel)]=\"fieldInfo.align\">\r\n                                            <option value=\"left\">\u5DE6\u5BF9\u9F50</option>\r\n                                            <option value=\"center\">\u5C45\u4E2D\u5BF9\u9F50</option>\r\n                                            <option value=\"right\">\u53F3\u5BF9\u9F50</option>\r\n                                        </select>\r\n                                    </div>\r\n                                </div>\r\n                            </div>\r\n                            <div class=\"farris-group-wrap\">\r\n                                <div class=\"form-group farris-form-group\">\r\n                                    <label for=\"hpinput01\" class=\"col-form-label\">\r\n                                        <span class=\"farris-label-text\">\u5BBD\u5EA6</span>\r\n                                    </label>\r\n                                    <div class=\"farris-input-wrap\">\r\n                                        <input type=\"input\" class=\"form-control\" [(ngModel)]=\"fieldInfo.size\">\r\n                                    </div>\r\n                                </div>\r\n                            </div> -->\r\n                    </div>\r\n\r\n                </div>\r\n            </div>\r\n\r\n        </div>\r\n       \r\n        <div region=\"south\" class=\"south\" [class.preview-open]=\"isOpen\" #preview>\r\n\r\n            <div class=\"d-flex flex-column flex-fill\" style=\"height: 100%;\">\r\n                <div region=\"north\" class=\"north preview-title\" (click)=\"isOpen = !isOpen\">\r\n                   <span class=\"text\" >\u6548\u679C\u9884\u89C8</span> \r\n                   <div class=\"close-btn\" (click)=\"closePreview($event)\">\r\n                        <span class=\"f-icon f-icon-close\"></span>\r\n                   </div>\r\n                </div>\r\n                <div region=\"center\" class=\"center flex-fill preview-container\">\r\n                    <table>\r\n\r\n                        <tr  *ngFor=\"let cls of previewGroupColumns\" style=\"height: 36px\">\r\n                            <ng-container *ngFor=\"let c of cls\">\r\n                                <ng-container *ngIf=\"c.colspan > 1\">\r\n                                    <td [attr.colspan]=\"c.colspan\" [attr.align]=\"c.halign || 'left'\">{{ c.title }}</td>\r\n                                </ng-container>\r\n\r\n                                <ng-container *ngIf=\"c.colspan == 1\">\r\n                                    <td [style.width.px]=\"c.width\"  [attr.align]=\"c.halign || 'left'\" [attr.rowspan]=\"c.rowspan || 1\">{{ c.title }}</td>\r\n                                </ng-container>\r\n                            </ng-container>\r\n                        </tr>\r\n                        \r\n\r\n                    </table>\r\n                </div>\r\n            </div>\r\n\r\n        </div>\r\n       \r\n</div>\r\n\r\n<ng-template #dlgFooter>\r\n    <div class=\"btns\">\r\n        <button  type=\"button\"  class=\"btn btn-primary btn-lg px-3\" (click)=\"ok()\" > \u786E\u5B9A </button>\r\n        &nbsp;&nbsp;\r\n        <button type=\"button\" class=\"btn btn-secondary btn-lg px-3\" (click)=\"cancle()\" > \u53D6\u6D88 </button> \r\n    </div>\r\n</ng-template>\r\n\r\n<ng-template #code>\r\n<textarea name=\"group-columns-code\" class=\"form-control\" rows=17>\r\n{{ columnsCode | json }}\r\n</textarea>\r\n</ng-template>\r\n\r\n\r\n",
                styles: [`
        .center::-webkit-scrollbar { width: 5px; height: 5px}
        .south::-webkit-scrollbar { width: 5px; height: 5px }
        .south {
            height: 40px;
            min-height: 40px;
            overflow: hidden;
            border-top: 1px solid #c8c8c8;
            background: #F3F8FF;
            transition: height 0.5s;
            flex-shrink: 0!important;;
        }
        .preview-open {
            height: 300px;
        }
        .preview-container {
            flex-shrink: 0!important;overflow:auto
        }
        .preview-title {
            height:36px;min-height: 36px; background: #F3F8FF; cursor:pointer;
        }
        .preview-title .text {
            line-height: 36px;font-size: 16px;padding-left: 10px;
        }
        .preview-title button {
            float: right;
            line-height: 40px;
            margin-right: 10px;
        }
        .preview-container table{
            width: max-content; margin: 5px; background-color: #efefef; border-color: #adadad;
        }
        .preview-container table td{border: 1px solid #c2c2c2;}

        .east.property-column {
            width: 300px; min-width:260px;border-left: 1px solid #c8c8c8;overflow: auto;background:#F3F8FF
        }
        .close-btn {
            display: inline-block;
            float: right;
            line-height: 36px;
            margin-right: 10px;
            width: 36px;
            color: #ccc;
            text-align: center;
        }
        .close-btn:hover {
            color: red;
        }
        `]
            }] }
];
/** @nocollapse */
ColumnGroupEditorComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: MessagerService },
    { type: Injector },
    { type: BsModalService }
];
ColumnGroupEditorComponent.propDecorators = {
    editorParams: [{ type: Input }],
    fields: [{ type: Input }],
    value: [{ type: Input }],
    cls: [{ type: HostBinding, args: ['class',] }],
    closeModal: [{ type: Output }],
    submitModal: [{ type: Output }],
    hge: [{ type: ViewChild, args: ['hge',] }],
    modalFooter: [{ type: ViewChild, args: ['dlgFooter',] }],
    previewEl: [{ type: ViewChild, args: ['preview',] }],
    codeTemplate: [{ type: ViewChild, args: ['code',] }]
};
if (false) {
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.editorParams;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.fields;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.value;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.cls;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.closeModal;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.submitModal;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.hge;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.modalFooter;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.previewEl;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.codeTemplate;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.previewGroupColumns;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.columnsCode;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.btnTopActive;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.btnBottomActive;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.isOpen;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.fieldInfo;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.modalConfig;
    /**
     * @type {?}
     * @private
     */
    ColumnGroupEditorComponent.prototype.render;
    /**
     * @type {?}
     * @private
     */
    ColumnGroupEditorComponent.prototype.messagerSer;
    /** @type {?} */
    ColumnGroupEditorComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    ColumnGroupEditorComponent.prototype.modalSer;
}
