// @farris/header-group-editor/esm2015/lib/group-header-select-columns.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectorRef } from '@angular/core';
export class SelectColumnsComponent {
    /**
     * @param {?} cd
     */
    constructor(cd) {
        this.cd = cd;
        this.fields = [];
        this.checkeds = [];
        this.uncheckeds = [];
        this.childs = [];
        this.checkedPropertyName = '__checked__';
        // 候选列
        this.hxl = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.initCheckedProperty();
        this.childs = Array.from(this.checkeds);
        if (this.groupEditorRef) {
            this.hxl = this.groupEditorRef.getHouXuanLie();
            if (this.hxl.length) {
                this.hxl = this.hxl.map((/**
                 * @param {?} n
                 * @return {?}
                 */
                n => n.data));
            }
        }
    }
    /**
     * @private
     * @return {?}
     */
    initCheckedProperty() {
        if (!this.checkeds) {
            this.checkeds = [];
            this.fields.forEach((/**
             * @param {?} f
             * @return {?}
             */
            f => {
                f[this.checkedPropertyName] = false;
            }));
        }
        else {
            this.fields.forEach((/**
             * @param {?} f
             * @return {?}
             */
            f => {
                f[this.checkedPropertyName] = this.isChecked(f);
            }));
        }
    }
    /**
     * @param {?} f
     * @return {?}
     */
    checkedCols(f) {
        if (this.isDisabled(f)) {
            return;
        }
        /** @type {?} */
        const checked = this.isChecked(f);
        f[this.checkedPropertyName] = !checked;
        /** @type {?} */
        const _fields = this.fields.filter((/**
         * @param {?} n
         * @return {?}
         */
        n => !this.isDisabled(n)));
        this.checkeds = _fields.filter((/**
         * @param {?} n
         * @return {?}
         */
        n => n[this.checkedPropertyName]));
        this.uncheckeds = _fields.filter((/**
         * @param {?} n
         * @return {?}
         */
        n => !n[this.checkedPropertyName]));
        this.cd.detectChanges();
    }
    /**
     * @param {?} f
     * @return {?}
     */
    isChecked(f) {
        /** @type {?} */
        let chk = false;
        if (this.checkeds && this.checkeds.length) {
            chk = this.checkeds.findIndex((/**
             * @param {?} n
             * @return {?}
             */
            n => n.id === f.id)) > -1;
        }
        return chk;
    }
    /**
     * 在其他分组中的列，禁用选择
     * 仅当前分组中的列及候选列中的项可以选择
     * @param {?} f
     * @return {?}
     */
    isDisabled(f) {
        /** @type {?} */
        let items = this.childs;
        if (this.hxl && this.hxl.length) {
            items = items.concat(this.hxl);
        }
        return !(items.findIndex((/**
         * @param {?} n
         * @return {?}
         */
        n => n.id === f.id)) > -1);
    }
}
SelectColumnsComponent.decorators = [
    { type: Component, args: [{
                selector: 'select-columns',
                template: `
    <div style="margin: 10px 20px" class="row">
        <div class="custom-control custom-checkbox col-4" *ngFor="let f of fields">
            <input type="checkbox" class="custom-control-input" [checked]="isChecked(f)" [disabled]="isDisabled(f)">
            <label class="custom-control-label" (click)="checkedCols(f)" >{{ f.caption }}</label>
        </div>
    </div>
    `
            }] }
];
/** @nocollapse */
SelectColumnsComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
SelectColumnsComponent.propDecorators = {
    fields: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    SelectColumnsComponent.prototype.fields;
    /** @type {?} */
    SelectColumnsComponent.prototype.checkeds;
    /** @type {?} */
    SelectColumnsComponent.prototype.uncheckeds;
    /** @type {?} */
    SelectColumnsComponent.prototype.childs;
    /** @type {?} */
    SelectColumnsComponent.prototype.groupEditorRef;
    /**
     * @type {?}
     * @private
     */
    SelectColumnsComponent.prototype.checkedPropertyName;
    /**
     * @type {?}
     * @private
     */
    SelectColumnsComponent.prototype.hxl;
    /**
     * @type {?}
     * @private
     */
    SelectColumnsComponent.prototype.cd;
}
