// @farris/ui-datagrid/esm2015/lib/components/header/drag-drop-column.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class DragDropColumnService {
    constructor() { }
    /**
     * 元素位置调整
     * arr: 数组
     * newIndex: 新位置索引
     * removedIndex: 老位置索引
     * @param {?} arr
     * @param {?} newIndex
     * @param {?} removedIndex
     * @return {?}
     */
    moveItem(arr, newIndex, removedIndex) {
        /** @type {?} */
        const removedItem = arr.splice(removedIndex, 1);
        arr.splice(newIndex, 0, ...removedItem);
        return arr;
    }
    /**
     * @param {?} columns
     * @return {?}
     */
    backupColumns(columns) {
        this.initColumns = columns.map((/**
         * @param {?} cols
         * @return {?}
         */
        cols => {
            return cols.map((/**
             * @param {?} n
             * @return {?}
             */
            n => {
                return {
                    id: n.id,
                    field: n.field
                };
            }));
        }));
    }
}
DragDropColumnService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
DragDropColumnService.ctorParameters = () => [];
/** @nocollapse */ DragDropColumnService.ngInjectableDef = i0.defineInjectable({ factory: function DragDropColumnService_Factory() { return new DragDropColumnService(); }, token: DragDropColumnService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    DragDropColumnService.prototype.initColumns;
}
