// @edp-bif/runtime-api/esm5/lib/bill-tracker/components/basic-table/child-row-table.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/components/basic-table/child-row-table.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, Input, ViewContainerRef, ComponentFactoryResolver, Output, EventEmitter, ChangeDetectorRef } from '@angular/core';
import { BasicTableComponent } from './basic-table.component';
import { BillDataRow, BillDataTable, BillSchemaTable } from '@edp-aif/common-api';
import { BillInfo } from '../../../entity/tracking/BillInfo';
var ChildRowTableDirective = /** @class */ (function () {
    function ChildRowTableDirective(cfr, vcRef, cdr) {
        this.cfr = cfr;
        this.vcRef = vcRef;
        this.cdr = cdr;
        this.hasChildren = new EventEmitter();
    }
    Object.defineProperty(ChildRowTableDirective.prototype, "expanded", {
        get: /**
         * @return {?}
         */
        function () {
            return this._expanded;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._expanded = val;
            if (val) {
                this.showChildTable();
            }
            else {
                this.clearChildTable();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ChildRowTableDirective.prototype.showChildTable = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        if (this.parentRow == null) {
            return;
        }
        this.clearChildTable();
        /** @type {?} */
        var cmpF = this.cfr.resolveComponentFactory(BasicTableComponent);
        try {
            for (var _b = tslib_1.__values(this.parentRow.GetChildTableList()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var childTable = _c.value;
                /** @type {?} */
                var cmpR = this.vcRef.createComponent(cmpF);
                cmpR.instance.billInfo = this.billInfo;
                cmpR.instance.isMain = false;
                cmpR.instance.showTable(this.parentSchemaTable.GetChildTableMap()[childTable.tableName], childTable);
                this.vcRef.insert(cmpR.hostView);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.cdr.markForCheck();
    };
    /**
     * @return {?}
     */
    ChildRowTableDirective.prototype.clearChildTable = /**
     * @return {?}
     */
    function () {
        this.vcRef.clear();
    };
    ChildRowTableDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[appChildRowTable]'
                },] }
    ];
    /** @nocollapse */
    ChildRowTableDirective.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: ViewContainerRef },
        { type: ChangeDetectorRef }
    ]; };
    ChildRowTableDirective.propDecorators = {
        billInfo: [{ type: Input, args: ["billInfo",] }],
        colspan: [{ type: Input, args: ["colspan",] }],
        parentRow: [{ type: Input, args: ["parentRow",] }],
        parentSchemaTable: [{ type: Input, args: ["parentSchemaTable",] }],
        parentDataTable: [{ type: Input, args: ["parentDataTable",] }],
        expanded: [{ type: Input, args: ["expanded",] }],
        hasChildren: [{ type: Output, args: ["hasChildren",] }]
    };
    return ChildRowTableDirective;
}());
export { ChildRowTableDirective };
if (false) {
    /** @type {?} */
    ChildRowTableDirective.prototype.billInfo;
    /** @type {?} */
    ChildRowTableDirective.prototype.colspan;
    /** @type {?} */
    ChildRowTableDirective.prototype.parentRow;
    /** @type {?} */
    ChildRowTableDirective.prototype.parentSchemaTable;
    /** @type {?} */
    ChildRowTableDirective.prototype.parentDataTable;
    /**
     * @type {?}
     * @private
     */
    ChildRowTableDirective.prototype._expanded;
    /** @type {?} */
    ChildRowTableDirective.prototype.hasChildren;
    /**
     * @type {?}
     * @private
     */
    ChildRowTableDirective.prototype.cfr;
    /** @type {?} */
    ChildRowTableDirective.prototype.vcRef;
    /** @type {?} */
    ChildRowTableDirective.prototype.cdr;
}
