// @edp-bif/runtime-api/esm2015/lib/bill-tracker/components/basic-table/child-row-table.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bill-tracker/components/basic-table/child-row-table.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ViewContainerRef, ComponentFactoryResolver, Output, EventEmitter, ChangeDetectorRef } from '@angular/core';
import { BasicTableComponent } from './basic-table.component';
import { BillDataRow, BillDataTable, BillSchemaTable } from '@edp-aif/common-api';
import { BillInfo } from '../../../entity/tracking/BillInfo';
export class ChildRowTableDirective {
    /**
     * @param {?} cfr
     * @param {?} vcRef
     * @param {?} cdr
     */
    constructor(cfr, vcRef, cdr) {
        this.cfr = cfr;
        this.vcRef = vcRef;
        this.cdr = cdr;
        this.hasChildren = new EventEmitter();
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set expanded(val) {
        this._expanded = val;
        if (val) {
            this.showChildTable();
        }
        else {
            this.clearChildTable();
        }
    }
    /**
     * @return {?}
     */
    get expanded() {
        return this._expanded;
    }
    /**
     * @return {?}
     */
    showChildTable() {
        if (this.parentRow == null) {
            return;
        }
        this.clearChildTable();
        /** @type {?} */
        const cmpF = this.cfr.resolveComponentFactory(BasicTableComponent);
        for (let childTable of this.parentRow.GetChildTableList()) {
            /** @type {?} */
            const cmpR = this.vcRef.createComponent(cmpF);
            cmpR.instance.billInfo = this.billInfo;
            cmpR.instance.isMain = false;
            cmpR.instance.showTable(this.parentSchemaTable.GetChildTableMap()[childTable.tableName], childTable);
            this.vcRef.insert(cmpR.hostView);
        }
        this.cdr.markForCheck();
    }
    /**
     * @return {?}
     */
    clearChildTable() {
        this.vcRef.clear();
    }
}
ChildRowTableDirective.decorators = [
    { type: Directive, args: [{
                selector: '[appChildRowTable]'
            },] }
];
/** @nocollapse */
ChildRowTableDirective.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: ViewContainerRef },
    { type: ChangeDetectorRef }
];
ChildRowTableDirective.propDecorators = {
    billInfo: [{ type: Input, args: ["billInfo",] }],
    colspan: [{ type: Input, args: ["colspan",] }],
    parentRow: [{ type: Input, args: ["parentRow",] }],
    parentSchemaTable: [{ type: Input, args: ["parentSchemaTable",] }],
    parentDataTable: [{ type: Input, args: ["parentDataTable",] }],
    expanded: [{ type: Input, args: ["expanded",] }],
    hasChildren: [{ type: Output, args: ["hasChildren",] }]
};
if (false) {
    /** @type {?} */
    ChildRowTableDirective.prototype.billInfo;
    /** @type {?} */
    ChildRowTableDirective.prototype.colspan;
    /** @type {?} */
    ChildRowTableDirective.prototype.parentRow;
    /** @type {?} */
    ChildRowTableDirective.prototype.parentSchemaTable;
    /** @type {?} */
    ChildRowTableDirective.prototype.parentDataTable;
    /**
     * @type {?}
     * @private
     */
    ChildRowTableDirective.prototype._expanded;
    /** @type {?} */
    ChildRowTableDirective.prototype.hasChildren;
    /**
     * @type {?}
     * @private
     */
    ChildRowTableDirective.prototype.cfr;
    /** @type {?} */
    ChildRowTableDirective.prototype.vcRef;
    /** @type {?} */
    ChildRowTableDirective.prototype.cdr;
}
