// @farris/ui-feature-editor/fesm2015/farris-ui-feature-editor.js
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { EventManager } from '@angular/platform-browser';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { FarrisFormsModule } from '@farris/ui-forms';
import { ComboListModule } from '@farris/ui-combo-list';
import { InputGroupModule } from '@farris/ui-input-group';
import { BsModalService, ModalModule } from '@farris/ui-modal';
import { FarrisDatePickerModule } from '@farris/ui-datepicker';
import { NumberSpinnerModule } from '@farris/ui-number-spinner';
import { TimePickerModule } from '@farris/ui-time-picker';
import { LocaleService, LocaleModule } from '@farris/ui-locale';
import { SwitchModule } from '@farris/ui-switch';
import { FarrisSectionModule } from '@farris/ui-section';
import { Injectable, Component, Input, Inject, LOCALE_ID, Optional, ComponentFactoryResolver, EventEmitter, Injector, Directive, Output, NgModule } from '@angular/core';
import { LookupModule } from '@farris/ui-lookup';
import { isObservable, of } from 'rxjs';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const InputType = {
    Help: 'Help',
    String: 'String',
    Enum: 'Enum',
    Date: 'Date',
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class LookupService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @param {?} url
     * @param {?} params
     * @return {?}
     */
    getData(url, params) {
        /** @type {?} */
        const urls = url.split(':');
        /** @type {?} */
        const realUrl = '/api/scm/scmfnd/v1.0/scmcommoncmpservice/service/gethelpdata';
        /** @type {?} */
        const helpID = urls[0] === '1' ? '9e2dca82-0aa2-4394-b26d-cb91164f44d5' : '344ffc53-7167-48d1-866b-a84083404e9e';
        /** @type {?} */
        const filter = 'setid:' + urls[1] + '&' + 'state_isenabled:1';
        /** @type {?} */
        const httpParams = {};
        if (params) {
            if (params.pageIndex) {
                httpParams['pageIndex'] = JSON.stringify(params.pageIndex - 0);
            }
            if (params.pageSize) {
                httpParams['pageSize'] = JSON.stringify(params.pageSize - 0);
            }
            if (params.condition) {
                httpParams['condition'] = JSON.stringify(params.condition);
            }
            if (params.searchValue) {
                httpParams['search'] = params.searchValue;
            }
        }
        return this.http.put(realUrl, {
            helpID: helpID,
            queryParam: JSON.stringify(httpParams),
            filterStr: filter
        }).pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        (data) => JSON.parse(data.returnValue))));
    }
}
LookupService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
LookupService.ctorParameters = () => [
    { type: HttpClient }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class FeatureEditorComponent {
    /**
     * @param {?} localeService
     */
    constructor(localeService) {
        this.localeService = localeService;
        if (this.localeService) {
            this.localeId = this.localeService.localeId;
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * 帮助选择数据后的回调
     * @param {?} event event
     * @return {?}
     */
    onSelectRows(event) {
        const { code = null, rows = [] } = event || {};
        if (code) {
            /** @type {?} */
            const item = this.features.find((/**
             * @param {?} feature
             * @return {?}
             */
            (feature) => feature.field === code));
            item.results = rows;
        }
    }
}
FeatureEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'farris-feature-editor',
                template: "<farris-section class=\"f-section-form f-section-in-mainsubcard\" [showHeader]=\"false\" [enableMaximize]=\"false\">\r\n  <div class=\"f-form-layout farris-form-controls-inline\" [class.farris-form-controls-inline]=\"true\">\r\n    <div class=\"col-12 col-md-6 col-xl-3 col-el-2\"  *ngFor=\"let feature of features;let row = index\">\r\n      <div class=\"farris-group-wrap\">\r\n        <div class=\"form-group farris-form-group\">\r\n          <label class=\"col-form-label\" title=\"feature?.title\" for=\"{{feature?.id}}\"><span\r\n              class=\"farris-label-info text-danger\" *ngIf=\"feature?.options.required\">*</span><span\r\n              class=\"farris-label-text\">{{feature?.title}}</span></label>\r\n          <div class=\"farris-input-wrap\">\r\n            <!--\u5E2E\u52A9-->\r\n          <ng-template [ngIf]=\"feature?.editor?.controlType==='lookup'\">\r\n            <!-- <ng-container *ngTemplateOutlet=\"lookupTemplate;context:input\"></ng-container> -->\r\n            <farris-lookup-grid #lookupGrid [lookup-extra]=\"feature?.field\" [ins]=\"lookupGrid\" [enableClear]='feature?.options?.enableClear || false'\r\n              (onSelectRows)=\"onSelectRows($event)\" input-end-edit [uri]=\"feature?.options?.uri\" [editable]=\"feature?.options?.editable || false\"\r\n              [mapFields]=\"feature?.options?.mapFields\" [displayType]=\"feature?.options?.displayType || 'List'\"\r\n              [idField]=\"feature?.options?.idField\" [singleSelect]=\"feature?.options?.singleSelect || true\"\r\n              [pageSize]=\"feature?.options?.pageSize\" [pageIndex]=\"1\" [pagination]=\"feature?.options?.pagination\"\r\n              [textField]=\"feature?.options?.textField\" [valueField]=\"feature?.options?.valueField\"\r\n              [title]=\"feature?.options?.dialogTitle\" [showMaxButton]=\"feature?.options?.showMaxButton\"\r\n              [showCloseButton]=\"feature?.options?.showCloseButton\" [resizable]=\"feature?.options?.resizable\"\r\n              [context]=\"feature?.options?.context\" [expandLevel]=\"feature?.options?.expandLevel\"\r\n              [isRecordSize]=\"feature?.options?.isRecordSize\" [enableFullTree]=\"feature?.options?.enableFullTree\"\r\n              [loadTreeDataType]=\"'default'\" [(ngModel)]=\"feature.value\" [bindingData]=\"feature?.options?.data\" [dictPicking]=\"feature?.options?.dictPicking\" [dictPicked]=\"feature?.options?.dictPicked\">\r\n            </farris-lookup-grid>\r\n          </ng-template>\r\n          <!--\u8F93\u5165\u6846-->\r\n          <ng-template\r\n            [ngIf]=\"feature?.editor?.controlType==='textbox' ||feature?.editor?.controlType==='textarea' || feature?.editor?.controlType === 'input-group'\">\r\n            <!-- <ng-container *ngTemplateOutlet=\"textboxTemplate;context:input\"></ng-container> -->\r\n            <input-group [(ngModel)]=\"feature.value\"></input-group>\r\n          </ng-template>\r\n          <!--\u65E5\u671F-->\r\n          <ng-template [ngIf]=\"feature?.editor?.controlType === 'datepicker'\">\r\n            <!-- <ng-container *ngTemplateOutlet=\"datepickerTemplate;context:input\"></ng-container> -->\r\n            <farris-datepicker [(ngModel)]=\"feature.value\" [dateRange]=\"feature?.options?.dateRange\">\r\n            </farris-datepicker>\r\n          </ng-template>\r\n          <!--\u6570\u5B57-->\r\n          <ng-template [ngIf]=\"feature?.editor?.controlType === 'numberbox'\">\r\n            <!-- <ng-container *ngTemplateOutlet=\"numberboxTemplate;context:input\"></ng-container> -->\r\n            <farris-number-spinner [(ngModel)]=\"feature.value\" [min]=\"feature?.options?.min\"\r\n              [max]=\"feature?.options?.max\" [precision]=\"feature?.options?.precision\" [step]=\"feature?.options?.step\">\r\n            </farris-number-spinner>\r\n          </ng-template>\r\n          <!--\u4E0B\u62C9-->\r\n          <ng-template\r\n            [ngIf]=\"feature?.editor?.controlType === 'combolist' || feature?.editor?.controlType === 'select'\">\r\n            <!-- <ng-container *ngTemplateOutlet=\"combolistTemplate;context:input\"></ng-container> -->\r\n            <farris-combo-list [(ngModel)]=\"feature.value\"\r\n              [idField]=\"feature?.options?.valueField || feature?.options?.idField || 'id'\"\r\n              [textField]=\"feature?.options?.textField ||'label'\" [data]=\"feature?.options?.data\">\r\n            </farris-combo-list>\r\n          </ng-template>\r\n          <!--\u65F6\u95F4-->\r\n          <ng-template [ngIf]=\"feature?.editor?.controlType === 'timepicker'\">\r\n            <!-- <ng-container *ngTemplateOutlet=\"timepickerTemplate;context:input\"></ng-container> -->\r\n            <farris-time-picker [(ngModel)]=\"feature.value\" [use12Hours]=\"feature?.options?.use12Hours\">\r\n            </farris-time-picker>\r\n          </ng-template>\r\n          <!--\u590D\u9009\u6846-->\r\n          <ng-template [ngIf]=\"feature?.editor?.controlType === 'checkbox'\">\r\n            <div class=\"custom-control custom-checkbox f-checkradio-single pl-5\">\r\n              <input [id]=\"feature.id\" class=\"custom-control-input\" type=\"checkbox\" [(ngModel)]=\"feature.value\">\r\n              <label class=\"custom-control-label\" [for]=\"feature.id\"></label>\r\n            </div>\r\n          </ng-template>\r\n          <!--\u5F00\u5173-->\r\n          <ng-template [ngIf]=\"feature?.editor?.controlType === 'switch'\">\r\n            <farris-switch [(ngModel)]=\"feature.value\" class=\"ml-5\">\r\n            </farris-switch>\r\n          </ng-template>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</farris-section>",
                styles: ['feature-editor.component.css']
            }] }
];
/** @nocollapse */
FeatureEditorComponent.ctorParameters = () => [
    { type: LocaleService }
];
FeatureEditorComponent.propDecorators = {
    features: [{ type: Input }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class I18nService {
    /**
     * @param {?} localeId
     */
    constructor(localeId) {
        this.localeId = localeId;
    }
    /**
     * @return {?}
     */
    get title() {
        /** @type {?} */
        let result = '特性编辑器';
        switch (this.localeId) {
            case 'en':
                result = 'Feature Editor';
                break;
            case 'zh-CHT':
                result = '特性編輯器';
                break;
            default: break;
        }
        return result;
    }
}
I18nService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
I18nService.ctorParameters = () => [
    { type: String, decorators: [{ type: Optional }, { type: Inject, args: [LOCALE_ID,] }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class FeatureEditorService {
    /**
     * @param {?} injector
     * @param {?} modalService
     * @param {?} componentFactoryResolver
     * @param {?} localeService
     * @param {?} languageService
     */
    constructor(injector, modalService, componentFactoryResolver, localeService, languageService) {
        this.injector = injector;
        this.modalService = modalService;
        this.componentFactoryResolver = componentFactoryResolver;
        this.localeService = localeService;
        this.languageService = languageService;
        this.dialog = null;
        this.componentRef = null;
        this.keyDownHandler = null;
        this.onUserConfirmed = new EventEmitter();
        if (!this.languageService) {
            /** @type {?} */
            const localeId = this.injector.get(LOCALE_ID, 'zh-CHS');
            this.languageService = new I18nService(localeId);
        }
        this.eventManager = this.injector.get(EventManager, null);
    }
    /**
     * 弹出特性编辑器
     * @param {?} features 特征数组
     * @param {?=} options 弹窗配置
     * @return {?}
     */
    show(features, options) {
        if (!features || features.length < 1) {
            return;
        }
        this.canOpen(options).subscribe((/**
         * @param {?} canOpen
         * @return {?}
         */
        (canOpen) => {
            if (!canOpen) {
                return;
            }
            /** @type {?} */
            const componentFactory = this.componentFactoryResolver.resolveComponentFactory(FeatureEditorComponent);
            this.componentRef = componentFactory.create(this.injector);
            this.componentRef.instance.features = features;
            /** @type {?} */
            const titleText = this.languageService.title;
            /** @type {?} */
            const okText = this.localeService.getValue('batchEditDialog.okText');
            /** @type {?} */
            const cancelText = this.localeService.getValue('batchEditDialog.cancelText');
            const { width = 470, height = 500, title = titleText } = options || {};
            /** @type {?} */
            const modalOptions = {
                title,
                width,
                height,
                buttons: [
                    {
                        text: cancelText, cls: 'btn btn-secondary', handle: (/**
                         * @return {?}
                         */
                        () => {
                            this.dialog.close();
                        })
                    },
                    {
                        text: okText, cls: 'btn btn-primary',
                        handle: (/**
                         * @return {?}
                         */
                        () => {
                            this.onConfirmClick(options);
                        })
                    }
                ],
                showButtons: true,
                areaResponse: true,
                opened: (/**
                 * @param {?} e
                 * @return {?}
                 */
                (e) => {
                    this.registerShortcutKey(options, e.instance);
                }),
                closed: (/**
                 * @return {?}
                 */
                () => {
                    if (this.keyDownHandler) {
                        this.keyDownHandler();
                        this.keyDownHandler = null;
                    }
                })
            };
            this.dialog = this.modalService.show(this.componentRef, modalOptions);
        }));
    }
    /**
     * @private
     * @param {?} options
     * @param {?} dialogRef
     * @return {?}
     */
    registerShortcutKey(options, dialogRef) {
        // 回车，与确定按钮处理逻辑一至。
        /** @type {?} */
        const dlgContainerDom = dialogRef.dialog.location.nativeElement;
        if (dlgContainerDom && !this.keyDownHandler) {
            this.keyDownHandler = this.eventManager.addEventListener(dlgContainerDom, 'keydown', (/**
             * @param {?} e
             * @return {?}
             */
            (e) => {
                e.stopPropagation();
                if (e.key === 'Enter') {
                    // if (e.target['nodeName'] === 'INPUT' && !e.ctrlKey) {
                    //     return;
                    // }
                    this.onConfirmClick(options);
                }
                else if (e.key === 'Escape') {
                    this.dialog.close();
                }
                else if (e.key === 'Tab') {
                    this.onTabKeydownHandler(e.target);
                }
            }));
        }
        if (this.componentRef) {
            this.setFirstInputFocus();
        }
    }
    /**
     * @private
     * @return {?}
     */
    setFirstInputFocus() {
        /** @type {?} */
        const firstInputEl = this.componentRef.location.nativeElement.querySelector('input,textarea');
        if (firstInputEl) {
            firstInputEl.focus();
        }
    }
    /**
     * @private
     * @param {?} currentTarget
     * @return {?}
     */
    onTabKeydownHandler(currentTarget) {
        /** @type {?} */
        let allInputs = this.componentRef.location.nativeElement.querySelectorAll('input,textarea');
        allInputs = Array.from(allInputs);
        /** @type {?} */
        const currentInputIndex = allInputs.findIndex((/**
         * @param {?} n
         * @return {?}
         */
        n => n === currentTarget));
        /** @type {?} */
        let nextInputIndex = currentInputIndex + 1;
        if (nextInputIndex >= allInputs.length) {
            nextInputIndex = 0;
        }
        /** @type {?} */
        const nextInputTarget = allInputs[nextInputIndex];
        if (nextInputTarget) {
            setTimeout((/**
             * @return {?}
             */
            () => {
                nextInputTarget.focus();
            }));
        }
    }
    /**
     * 用户点击确认按钮
     * @private
     * @param {?=} options
     * @return {?}
     */
    onConfirmClick(options) {
        this.canClose(options).subscribe((/**
         * @param {?} result
         * @return {?}
         */
        (result) => {
            if (result === true) {
                this.onConfirm(options);
            }
        }));
    }
    /**
     * 确认
     * @private
     * @param {?=} options options
     * @return {?}
     */
    onConfirm(options) {
        /** @type {?} */
        const features = this.componentRef && this.componentRef.instance && this.componentRef.instance.features || [];
        this.dialog.close();
        if (options.onConfirm && typeof options.onConfirm) {
            options.onConfirm(features);
        }
        this.onUserConfirmed.next(features);
    }
    /**
     * 是否可以关闭
     * @private
     * @param {?=} options options
     * @return {?}
     */
    canClose(options) {
        /** @type {?} */
        const features = this.componentRef && this.componentRef.instance && this.componentRef.instance.features || [];
        /** @type {?} */
        const onConfirmingRef = options && options.onConfirming || null;
        if (!onConfirmingRef) {
            return of(true);
        }
        // 有确认前事件
        /** @type {?} */
        const onConfirming = onConfirmingRef(features);
        if (isObservable(onConfirming)) {
            return onConfirming;
        }
        else {
            return of(onConfirming);
        }
    }
    /**
     * 是否可以弹出特性编辑器
     * @private
     * @param {?=} options options
     * @return {?}
     */
    canOpen(options) {
        /** @type {?} */
        const onOpeningHandlerRef = options && options.onOpening || null;
        if (!onOpeningHandlerRef) {
            return of(true);
        }
        /** @type {?} */
        const onOpeningHandler = onOpeningHandlerRef(options);
        if (isObservable(onOpeningHandler)) {
            return onOpeningHandler;
        }
        else {
            return of(onOpeningHandler);
        }
    }
}
FeatureEditorService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FeatureEditorService.ctorParameters = () => [
    { type: Injector },
    { type: BsModalService },
    { type: ComponentFactoryResolver },
    { type: LocaleService },
    { type: I18nService }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class LookupDirective {
    constructor() {
        this.onSelectRows = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.lookup) {
            this.lookup.dictPicked = (/**
             * @param {?} rows
             * @return {?}
             */
            (rows) => {
                if (this.onSelectRows) {
                    this.onSelectRows.emit({ code: this.code, rows: [].concat(rows) });
                }
                /** @type {?} */
                const result = {
                    closeDialog: true
                };
                return of(result);
            });
        }
    }
}
LookupDirective.decorators = [
    { type: Directive, args: [{
                selector: '[lookup-extra]'
            },] }
];
/** @nocollapse */
LookupDirective.ctorParameters = () => [];
LookupDirective.propDecorators = {
    code: [{ type: Input, args: ['lookup-extra',] }],
    lookup: [{ type: Input, args: ['ins',] }],
    onSelectRows: [{ type: Output }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class FeatureEditorModule {
}
FeatureEditorModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    FeatureEditorComponent,
                    LookupDirective
                ],
                imports: [
                    CommonModule,
                    FormsModule,
                    FarrisFormsModule,
                    LookupModule,
                    ComboListModule,
                    InputGroupModule,
                    ModalModule,
                    FarrisDatePickerModule,
                    NumberSpinnerModule,
                    TimePickerModule,
                    SwitchModule,
                    FarrisSectionModule,
                    LocaleModule.forRoot(),
                ],
                exports: [
                    FeatureEditorComponent,
                    LookupDirective
                ],
                entryComponents: [
                    FeatureEditorComponent
                ],
                providers: [
                    I18nService
                ]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { InputType, LookupService, FeatureEditorService, I18nService, FeatureEditorComponent, FeatureEditorModule, LookupDirective as ɵa };
