// @farris/ui-list-filter/esm2015/lib/service/solution.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class SolutionService {
    constructor() {
        this.filterEditors = {};
    }
    /**
     * @return {?}
     */
    getGuid() {
        return (this.s4() + this.s4() + "-" + this.s4() + "-" + this.s4() + "-" + this.s4() + "-" + this.s4() + this.s4() + this.s4());
    }
    /**
     * @private
     * @return {?}
     */
    s4() {
        return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
    }
    /**
     * 根据字段获取Lable控件的Title
     * @param {?} field
     * @return {?}
     */
    getConditionItemTitle(field) {
        if (field.id && field.id !== '') {
            return field.fieldName;
        }
    }
    /**
     * 数组克隆 也可使用JSON.Parse(JSON.Stringify([input])),前提类结构中没有借口
     * @param {?} input
     * @return {?}
     */
    arrayClone(input) {
        /** @type {?} */
        let result = new Array();
        if (input && input.length > 0) {
            input.forEach((/**
             * @param {?} element
             * @return {?}
             */
            element => {
                result.push(Object.assign({}, element));
            }));
        }
        return result;
    }
    /**
     * @param {?} fieldName
     * @param {?} cmpIns
     * @return {?}
     */
    addFilterEditorInstance(fieldName, cmpIns) {
        this.filterEditors[fieldName] = cmpIns;
    }
}
SolutionService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
SolutionService.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    SolutionService.prototype.filterEditors;
}
