// @gsp-cmp/querysolution/esm2015/lib/query-solution.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/query-solution.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewEncapsulation, HostBinding, ComponentFactoryResolver, Injector, Input, Output, EventEmitter, ViewChild, InjectFlags, ElementRef, ChangeDetectorRef } from '@angular/core';
import { SaveAsDialogComponent } from './saveasdialog/saveasdialog.component';
import { SolutionManangerComponent } from './solutionmanager/solutionmanager.component';
import { Observable } from 'rxjs/Observable';
import { ConvertorService, QueryConditionComponent, QuerySolutionVo } from '@farris/component-querycondition';
import { NotifyService } from '@farris/ui-notify';
import { BsModalService } from '@farris/ui-modal';
import { HttpService } from './service/httpservice';
import { FieldConfigService } from './service/fieldconfig.service';
import { QuerySolutionVoConvertor, SolutionService } from '@farris/component-querycondition';
import { QUERYSOLUTION_HANDLER_TOKEN } from './interface/IQuerySolutionHandler';
import { LocaleService } from '@farris/ui-locale';
import { cloneDeep } from 'lodash-es';
import { SolutionCacheService } from './service/solution.cache.service';
import { EventManager } from '@angular/platform-browser';
import { IdService } from '@farris/ui-common';
import { BehaviorSubject, of, Subject } from 'rxjs';
import { switchMap, takeUntil, tap } from 'rxjs/operators';
import { EXTEND_QUERYSOLUTION_LOCALE } from './utils/extend-locale';
import { MessagerService } from '@farris/ui-messager';
export class QuerySolutionComponent {
    /**
     * @param {?} notify
     * @param {?} resolver
     * @param {?} injector
     * @param {?} modalService
     * @param {?} convertService
     * @param {?} solutionService
     * @param {?} httpService
     * @param {?} fieldConfigService
     * @param {?} localeService
     */
    constructor(notify, resolver, injector, modalService, convertService, solutionService, httpService, fieldConfigService, localeService) {
        this.notify = notify;
        this.resolver = resolver;
        this.injector = injector;
        this.modalService = modalService;
        this.convertService = convertService;
        this.solutionService = solutionService;
        this.httpService = httpService;
        this.fieldConfigService = fieldConfigService;
        this.localeService = localeService;
        this._fieldConfigs = [];
        this._presetFieldConfigs = [];
        this._resetFieldConfigs = [];
        /**
         * 高级模式需要原始fieldConfigs，需要将范围类控件改为单值控件
         */
        this.unformatFieldConfigs = [];
        /**
         * 此值由设计器传入
         * 组件在设计时无法点击被禁用，true
         * 组件在运行时可以点击使用，false
         */
        this.isDisabled = false;
        /**
         * 查询条件列表
         */
        this.queryEmitter = new EventEmitter();
        /**
         * 查询方案cls
         */
        this.cls = 'query-solution';
        /**
         * 查询方案数据源
         */
        this.querySolutionVos = new Array();
        this.querySolutionAdvancedVos = new Array();
        /**
         * 当前所选的方案，有可能不是方案
         */
        this.selectedSolution = new QuerySolutionVo();
        /**
         * 是否折叠
         */
        this.isFold = false;
        /**
         * label和控件是否排成一行，默认是；false时排成上下两行
         */
        this.isControlInline = true;
        /**
         * 是否需要完整显示超过6个字符的标签，默认false
         */
        this.showCompleteLabel = false;
        /**
         * 是否启用加载后查询
         */
        this.enableInitQuery = false;
        /**
         * 选择的控件展示列表
         */
        this.selectedFieldConditions = new Array();
        /**
         * visible属性为true的过滤条件，是变量selectedFieldConditions的子级
         */
        this.visibleSelectedConditions = new Array();
        //方案是否赋默认值
        this.isSetDefaultValue = false;
        //方案是否已赋默认值
        this.doneSetDefaultValue = false;
        //是否已完成自动查询，默认值为true，当enableInitQuery输入值为true时，首次执行ngOnchanges()时将其设为false
        this.doneInitQuery = true;
        //字段显示隐藏动态状态
        this.conditionVisibleState = new BehaviorSubject([]);
        //表单关闭时停止监听BehaviorSubject相关
        this.destroy$ = new Subject();
        /**
         * 条件概要
         */
        this.summaryConditions = [];
        /**
         * 是否启用记录历史方式
         * 上次查询——不能删除、可以另存、不存在变更
         * 方案列表只有上次查询时，上次查询、系统默认都存在
         * 方案列表多个时，有上次查询，上次查询选中
         * 方案列表没有上次查询时，但是LocalStorage有,构造一个上次查询
         *
         */
        this.enableHistory = false;
        /**
         * 是否启动高级模式
         */
        this.openAdvancedMode = false;
        this.filterText = this.localeService.getValue("queryCondition.container.query");
        this.queryAfterValueChange = false;
        this.hideOrgselector = false;
        this.historyRefer = {
            schemeKey: '',
            // 历史记录scheme
            foundSchme: null,
            // 历史方案保存
            schemeKeyLS: '',
            //LocalStorage 历史记录标记
            initCondition: false //LocalStorage是否按照执行完查询
        };
        this.newSaveAsSolutionVo = (/**
         * @return {?}
         */
        () => {
            //当前选中的方案，再修改筛选条件后，其原来的筛选条件并没有变化，仅仅this.selectedFieldConditions有变化
            /** @type {?} */
            const solution = cloneDeep(this.selectedSolution);
            //新增的方案要清空查询方案编号
            solution.code = '';
            //是否系统属性必须设置为False
            solution.isSystem = false;
            //解决方案所属表单
            solution.belongId = this.formId;
            //新增方案Id需要Guid
            solution.id = this.solutionService.getGuid();
            //新增方案isDefault需要设置为false
            solution.isDefault = false;
            //新增方案的筛选条件，是当前选中的筛选条件
            solution.queryConditions = this.selectedFieldConditions;
            //设置类型要清空
            solution.type = "private";
            solution.isChanged = false;
            solution.extendId = 'query-scheme-1';
            solution.extendInfo = new Array();
            return solution;
        });
        this.handleKeyDown = (/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            if (event.key == 'Enter') {
                //高级模式下数据格式的问题，输入状态中直接回车，输入框类型控件无法更新value，手动触发一下失去焦点，更新状态
                if (this.selectedSolution.mode === '2') {
                    /** @type {?} */
                    let activeEle = (/** @type {?} */ (document.activeElement));
                    activeEle.blur();
                    activeEle.focus();
                }
                this.query();
            }
        });
        this._idSer = new IdService();
        this.solutionHandler = this.injector.get(QUERYSOLUTION_HANDLER_TOKEN, null, InjectFlags.Optional);
        this.msgSer = this.injector.get(MessagerService);
        this._self = this.injector.get(ElementRef);
        this.cdf = this.injector.get(ChangeDetectorRef);
        this.eventManager = this.injector.get(EventManager);
        if (!this.localeService.langData['extendQuerySolution']) {
            this.localeService.setLocaleData(EXTEND_QUERYSOLUTION_LOCALE);
        }
    }
    /**
     * 字段配置数据源，用于选择字段配置
     * @param {?} fieldConfigJson
     * @return {?}
     */
    set fieldConfigs(fieldConfigJson) {
        this._fieldConfigs = this.solutionService.convertToQueryConditions(fieldConfigJson);
        this.unformatFieldConfigs = fieldConfigJson;
    }
    /**
     * @return {?}
     */
    get fieldConfigs() {
        return this._fieldConfigs;
    }
    /**
     * @param {?} resetFieldConfigJson
     * @return {?}
     */
    set resetFieldConfigs(resetFieldConfigJson) {
        if (!resetFieldConfigJson || !Array.isArray(resetFieldConfigJson) || resetFieldConfigJson.length == 0) {
            return;
        }
        // 规范数据
        /** @type {?} */
        let formatResetConfigJson = resetFieldConfigJson.map((/**
         * @param {?} tConfigItem
         * @return {?}
         */
        tConfigItem => {
            /** @type {?} */
            let findItem = this.fieldConfigs.find((/**
             * @param {?} presetItem
             * @return {?}
             */
            presetItem => presetItem['fieldCode'] == tConfigItem['labelCode']));
            if (findItem) {
                if (!tConfigItem.hasOwnProperty('control')) {
                    tConfigItem['control'] = {};
                }
                tConfigItem['control']['controltype'] = findItem.control.getControlType();
                return tConfigItem;
            }
        }));
        this._resetFieldConfigs = this.solutionService.convertToSimpleQueryConditions(formatResetConfigJson.filter((/**
         * @param {?} item
         * @return {?}
         */
        item => item)));
        if (this._resetFieldConfigs.length > 0) {
            this.updateConditionVisibleState(this._resetFieldConfigs);
        }
        // 更新已有配置
        this.solutionService.updateControl(this._presetFieldConfigs, this._resetFieldConfigs);
        this.solutionService.updateControl(this._fieldConfigs, this._resetFieldConfigs);
        this._fieldConfigs = this._fieldConfigs.slice();
        // 更新已有的方案
        this.sysQuerySolutionVo && this.solutionService.updateControl(this.sysQuerySolutionVo.queryConditions, this._resetFieldConfigs);
        // 更新历史记录
        this.historyRefer['foundSchme'] && this.solutionService.updateControl(this.historyRefer['foundSchme'].queryConditions, this._resetFieldConfigs);
        // 更新已选
        this.selectedSolution && this.solutionService.updateControl(this.selectedSolution.queryConditions, this._resetFieldConfigs);
    }
    /**
     * @return {?}
     */
    get resetFieldConfigs() {
        return this._resetFieldConfigs;
    }
    /**
     * 预置查询字段
     * @param {?} presetFieldConfigJson
     * @return {?}
     */
    set presetFieldConfigs(presetFieldConfigJson) {
        this._presetFieldConfigs = this.solutionService.convertToQueryConditions(presetFieldConfigJson);
    }
    /**
     * @return {?}
     */
    get presetFieldConfigs() {
        return this._presetFieldConfigs;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._self.nativeElement.removeEventListener("keydown", this.handleKeyDown);
        //this.removeCurrentSolutionCache();
        this.destroy$.next();
        this.destroy$.complete();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes['defaultConfigs']) {
            this.isSetDefaultValue = true;
            if (changes['defaultConfigs'].currentValue && !changes['defaultConfigs'].firstChange) { //设计时不会给defaultConfigs赋值
                this.setQueryConditionValues(changes['defaultConfigs'].currentValue, this.selectedSolution);
                this.initAutoQuery(this.selectedSolution);
            }
        }
        if (changes['enableInitQuery'] && changes['enableInitQuery'].firstChange && changes['enableInitQuery'].currentValue) {
            this.doneInitQuery = false;
        }
    }
    /**
     * 组件初始化
     * @return {?}
     */
    ngOnInit() {
        //只在运行时 获取sessionId，查询已存储方案
        this.initSysQuerySolution();
        if (!this.isDisabled) {
            this.sessionId = this.solutionHandler.getUserSessionId();
            // 缓存历史方案
            this.initAutoQueryByLS();
            // 加载方案列表
            this.loadQuerySolutions();
            //注册键盘监听事件
            this.listenKeyBoard();
            this.getConditionVisibleState().pipe(takeUntil(this.destroy$)).subscribe((/**
             * @param {?} simpleConditions
             * @return {?}
             */
            (simpleConditions) => {
                if (this.selectedFieldConditions.length > 0) {
                    this.updateFieldParams(simpleConditions);
                }
                else {
                    this.templateSimpleConditions = simpleConditions;
                }
                this.visibleSelectedConditions = this.selectedFieldConditions.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                (item) => { return !!item.visible; }));
                this.whenFoldSetSummaryConditions();
                !this.isDisabled && this.solutionHandler.setCurrentQueryConditions(cloneDeep(this.visibleSelectedConditions));
            }));
        }
        else {
            //设计时只加载系统预置方案
            this.loadSysQuerySolution();
        }
    }
    /**
     * @return {?}
     */
    initSysQuerySolution() {
        if (this.presetQuerySolutionName && this.presetQuerySolutionName != '') {
            this.sysQuerySolutionVo = new QuerySolutionVo();
            this.sysQuerySolutionVo.id = this.solutionService.getGuid();
            this.sysQuerySolutionVo.code = (this.presetQuerySolutionName != this.localeService.getValue("querySolution.container.sysPresetName")) ? this.presetQuerySolutionName : this.localeService.getValue("querySolution.container.defaultName");
            this.sysQuerySolutionVo.queryConditions = cloneDeep(this.presetFieldConfigs);
            this.sysQuerySolutionVo.belongId = this.formId;
            this.sysQuerySolutionVo.isSystem = true;
            this.sysQuerySolutionVo.mode = '1';
            // this.querySolutionVos.push(this.sysQuerySolutionVo);
        }
    }
    /**
     * 加载系统预置方案
     * @return {?}
     */
    loadSysQuerySolution() {
        this.querySolutionVos.push(this.sysQuerySolutionVo);
        this.selectedSolution = cloneDeep(this.sysQuerySolutionVo);
        this.loadQueryConditions();
    }
    /**
     * 规范方案条件
     * @private
     * @param {?} tConvertor
     * @param {?} element
     * @param {?=} fromJobject
     * @return {?}
     */
    formatQuerySolutionVO(tConvertor, element, fromJobject = false) {
        /** @type {?} */
        let querysolutionVO;
        /** @type {?} */
        let sourceData;
        /** @type {?} */
        let _fieldCodeArr = [];
        /** @type {?} */
        let _tQueryConditions;
        if (fromJobject) {
            querysolutionVO = tConvertor.initFromJobject(element);
        }
        else {
            querysolutionVO = element;
        }
        this.fieldConfigs.forEach((/**
         * @param {?} _item
         * @return {?}
         */
        _item => { _fieldCodeArr.push(_item.fieldCode); }));
        if (querysolutionVO && querysolutionVO.queryConditions && querysolutionVO.queryConditions.length > 0) {
            //若表单上删除了某个字段，但已保存的自定义方案中包含该字段时，在加载方案过程中将该字段剔除
            _tQueryConditions = querysolutionVO.queryConditions.filter((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                return _fieldCodeArr.indexOf(element.fieldCode) > -1;
            }));
            querysolutionVO.queryConditions = _tQueryConditions;
            //对自定义方案的国际化信息、帮助控件的帮助前及帮助后事件进行动态赋值
            querysolutionVO.queryConditions.forEach((/**
             * @param {?} element
             * @return {?}
             */
            element => {
                sourceData = this.fieldConfigs.find((/**
                 * @param {?} _item
                 * @return {?}
                 */
                _item => _item.fieldCode == element.fieldCode));
                element.fieldName = sourceData.fieldName;
                element.placeHolder = sourceData.placeHolder;
                element.beginPlaceHolder = sourceData.hasOwnProperty('beginPlaceHolder') ? sourceData['beginPlaceHolder'] : '';
                element.endPlaceHolder = sourceData.hasOwnProperty('endPlaceHolder') ? sourceData['endPlaceHolder'] : '';
                if (sourceData.control.preEventCmd) {
                    ((/** @type {?} */ (element.control))).preEventCmd = sourceData.control.preEventCmd;
                }
                if (sourceData.control.postEventCmd) {
                    ((/** @type {?} */ (element.control))).postEventCmd = sourceData.control.postEventCmd;
                }
                if (sourceData.control.clear) {
                    ((/** @type {?} */ (element.control))).clear = sourceData.control.clear;
                }
                if (sourceData.control.click) {
                    ((/** @type {?} */ (element.control))).click = sourceData.control.click;
                }
                if (sourceData.control.beforeShow) {
                    ((/** @type {?} */ (element.control))).beforeShow = sourceData.control.beforeShow;
                }
                if (sourceData.control.beforeHide) {
                    ((/** @type {?} */ (element.control))).beforeHide = sourceData.control.beforeHide;
                }
                if (sourceData.control.dateChangedCmd) {
                    ((/** @type {?} */ (element.control))).dateChangedCmd = sourceData.control.dateChangedCmd;
                }
                if (sourceData.control.enumValues && sourceData.control.enumValues.length > 0) {
                    ((/** @type {?} */ (element.control))).enumValues = sourceData.control.enumValues;
                    ((/** @type {?} */ (element.control))).enumValues = sourceData.control.enumValues;
                }
                if (sourceData.control.data && sourceData.control.data.length > 0) {
                    ((/** @type {?} */ (element.control))).data = sourceData.control.data;
                }
            }));
            //在此处querycondition处理完毕，需要将querycondition内容填充至queryConditionsAdvanced
            if (querysolutionVO.queryConditionsAdvanced) {
                /** @type {?} */
                let _queryConditionsAdvanced = tConvertor.complicateTreeData(querysolutionVO.queryConditionsAdvanced, querysolutionVO.queryConditions);
                querysolutionVO.queryConditionsAdvanced = cloneDeep(_queryConditionsAdvanced);
            }
        }
        return querysolutionVO;
    }
    /**
     * 加载查询方案数据
     * @return {?}
     */
    loadQuerySolutions() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /** @type {?} */
            const self = this;
            this.httpService.query(this.formId, this.sessionId).subscribe((/**
             * @param {?} items
             * @return {?}
             */
            (items) => {
                /** @type {?} */
                const convertor = new QuerySolutionVoConvertor();
                /** @type {?} */
                let querysolutionVO;
                if (items && items.length > 0) {
                    //高级模式未开启状态下，过滤掉相关数据
                    if (!this.openAdvancedMode) {
                        items = items.filter((/**
                         * @param {?} val
                         * @return {?}
                         */
                        val => val.mode !== '2'));
                    }
                    items.forEach((/**
                     * @param {?} element
                     * @return {?}
                     */
                    element => {
                        querysolutionVO = this.formatQuerySolutionVO(convertor, element, true);
                        if (this.enableHistory && querysolutionVO['type'] == 'pre' && querysolutionVO.mode !== '2') {
                            // 如果返回列表内存在历史记录                        
                            querysolutionVO.code = this.localeService.getValue("querySolution.container.histroyName");
                            this.historyRefer['schemeKey'] = querysolutionVO.id;
                            this.historyRefer['foundSchme'] = cloneDeep(querysolutionVO);
                            this.historyRefer['initCondition'] = true;
                            //  this.defaultQuerySolution = cloneDeep(querysolutionVO);
                        }
                        else if (querysolutionVO['type'] != 'pre') {
                            //根据version 选择存入哪一个数据 1为标准，2为高级
                            if (querysolutionVO.mode === '2') {
                                this.querySolutionAdvancedVos.push(querysolutionVO);
                            }
                            else {
                                this.querySolutionVos.push(querysolutionVO);
                            }
                            if (querysolutionVO.isDefault) {
                                //如果有默认方案，则设置全局默认方案
                                this.defaultQuerySolution = cloneDeep(querysolutionVO);
                            }
                        }
                    }));
                    //self.setSolutionCache(this.querySolutionVos);
                }
                if (this.enableHistory && !this.historyRefer['initCondition']) {
                    //如果还没有设置条件
                    this.formatInitQuerySolution();
                }
                if (this.enableHistory && this.historyRefer['initCondition']) {
                    this.selectedSolution = cloneDeep(this.historyRefer['foundSchme']);
                    //只有缓存记录
                    this.querySolutionVos.splice(0, 0, this.historyRefer['foundSchme']);
                    //历史记录不支持高级模式
                    // if(this.historyRefer['foundSchme'].mode === '2' ) {
                    //     this.querySolutionAdvancedVos.splice(0, 0, this.historyRefer['foundSchme']);
                    // } else {
                    //     this.querySolutionVos.splice(0, 0, this.historyRefer['foundSchme']);
                    // }
                    if (this.querySolutionVos.length === 1) {
                        // 只有上次查询时
                        this.querySolutionVos.push(this.sysQuerySolutionVo);
                    }
                    if (this.querySolutionAdvancedVos.length === 0) {
                        // 只有上次查询时
                        this.querySolutionAdvancedVos.push(this.initAdvancedSolution());
                    }
                    //self.setSolutionCache(this.querySolutionVos);
                    this.loadQueryConditions();
                    // if(this.historyRefer['schemeKey'] !== ''){
                    //     //更新一次缓存
                    //     this.changeQueryHistroy(false);
                    // }                
                }
                else if (this.defaultQuerySolution || this.querySolutionVos.length || this.querySolutionAdvancedVos.length) {
                    //如果已查到有默认方案，或已有存储的方案
                    //不再加载系统配置的方案，首先加载默认方案，没有默认则加载第一个方案
                    if (this.defaultQuerySolution && this.defaultQuerySolution.id != '') {
                        this.selectedSolution = cloneDeep(this.defaultQuerySolution);
                    }
                    else {
                        /** @type {?} */
                        const publicSolution = [...this.querySolutionVos, ...this.querySolutionAdvancedVos].find((/**
                         * @param {?} item
                         * @return {?}
                         */
                        item => item.type == 'public'));
                        if (publicSolution) {
                            this.selectedSolution = publicSolution;
                        }
                        else {
                            /** @type {?} */
                            const orgSolution = [...this.querySolutionVos, ...this.querySolutionAdvancedVos].find((/**
                             * @param {?} item
                             * @return {?}
                             */
                            item => item.type == 'org'));
                            this.selectedSolution = !!orgSolution ? orgSolution : cloneDeep([...this.querySolutionVos, ...this.querySolutionAdvancedVos][0]);
                        }
                    }
                    if (this.querySolutionVos.length < 1) {
                        this.querySolutionVos.push(this.sysQuerySolutionVo);
                    }
                    if (this.querySolutionAdvancedVos.length < 1) {
                        this.querySolutionAdvancedVos.push(this.initAdvancedSolution());
                    }
                    this.loadQueryConditions();
                }
                else {
                    //如果没有任何已存储的方案，则加载系统配置方案
                    this.loadSysQuerySolution();
                    this.querySolutionAdvancedVos.push(this.initAdvancedSolution());
                }
                if (this.isSetDefaultValue && !this.historyRefer['initCondition']) {
                    this.setQueryConditionValues(this.defaultConfigs, this.selectedSolution);
                }
                this.initAutoQuery(this.selectedSolution);
            }), (/**
             * @param {?} e
             * @return {?}
             */
            e => {
                if (e && e.message) {
                    this.msgSer.show('error', e.message, { showHeader: false });
                }
            }));
        });
    }
    /**
     * @param {?} querySolutionVos
     * @return {?}
     */
    setSolutionCache(querySolutionVos) {
        if (!this.isDisabled && gspframeworkService && querySolutionVos && querySolutionVos.length) {
            // 获取当前功能菜单信息, 并以功能菜单ID为缓存key
            /** @type {?} */
            const option = gspframeworkService.rtf.session.getCommonVariable();
            SolutionCacheService.setSolutionCache(option.tabId, querySolutionVos);
            SolutionCacheService.solutionListChanged.next(true);
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setCurrentSolutionCache(value) {
        if (!this.isDisabled && gspframeworkService && value) {
            // 获取当前功能菜单信息, 并以功能菜单ID为缓存key
            /** @type {?} */
            const option = gspframeworkService.rtf.session.getCommonVariable();
            if (option && option.tabId) {
                SolutionCacheService.setCurrentSolution(option.tabId, value);
            }
        }
    }
    /*
        removeCurrentSolutionCache() {
            if (!this.isDisabled && gspframeworkService) {
                // 获取当前功能菜单信息, 并以功能菜单ID为缓存key
                const option = gspframeworkService.rtf.session.getCommonVariable();
                if (option && option.tabId) {
                    SolutionCacheService.removeCurrentSolution(option.tabId);
                }
            }
        }
        */
    /**
     * @return {?}
     */
    loadQueryConditions() {
        //加载筛选条件
        this.selectedFieldConditions = new Array();
        if (this.selectedSolution && this.selectedSolution.queryConditions && this.selectedSolution.queryConditions.length > 0) {
            // this.selectedFieldConditions = this.selectedSolution.queryConditions as Array<QueryCondition>;
            this.selectedFieldConditions = this.selectedSolution.queryConditions;
            if (this.templateSimpleConditions && this.templateSimpleConditions.length > 0) {
                this.updateFieldParams(this.templateSimpleConditions);
                this.templateSimpleConditions = new Array();
            }
        }
        this.visibleSelectedConditions = this.selectedFieldConditions.filter((/**
         * @param {?} item
         * @return {?}
         */
        (item) => { return !!item.visible; }));
        // 如果方案的变更是在収折时
        this.whenFoldSetSummaryConditions();
        this.showSaveButton = this.selectedSolution.isSystem ? false : true;
        !this.isDisabled && this.solutionHandler.setCurrentQueryConditions(cloneDeep(this.visibleSelectedConditions));
        //this.setCurrentSolutionCache(this.selectedSolution);
    }
    /**
     * @return {?}
     */
    getConditionVisibleState() {
        return this.conditionVisibleState;
    }
    /**
     * 传递过来的是已经处理过的条件
     * @param {?} simpleCondition
     * @return {?}
     */
    updateConditionVisibleState(simpleCondition) {
        this.conditionVisibleState.next(simpleCondition);
    }
    // 监听resetFieldConfigs变化后，更新selectedFieldConditions，并同步至querycondition，可实现筛选条件动态显示隐藏
    /**
     * @private
     * @param {?} simpleConditions
     * @return {?}
     */
    updateFieldParams(simpleConditions) {
        if (!simpleConditions || !simpleConditions.length) {
            return;
        }
        /** @type {?} */
        const currentConditions = cloneDeep(this.selectedFieldConditions);
        /** @type {?} */
        const updatedConditions = currentConditions.map((/**
         * @param {?} item
         * @return {?}
         */
        (item) => {
            for (const simpleCondition of simpleConditions) {
                if (simpleCondition.hasOwnProperty('visible') && item.fieldCode == simpleCondition['fieldCode'] && item.visible !== simpleCondition['visible']) {
                    item.visible = simpleCondition['visible'];
                    item.value.clearValue();
                    this.markAsChanged(true, this.selectedSolution.id);
                }
            }
            return item;
        }));
        this.selectedFieldConditions = updatedConditions.slice();
    }
    /**
     * @return {?}
     */
    saveAs() {
        this.openSaveAsDialog(this.localeService.getValue("querySolution.saveAsDialog.title"));
    }
    /**
     * 获取默认标签
     * @param {?} solution
     * @return {?}
     */
    getDefaultLabel(solution) {
        if (!(this.defaultQuerySolution)) {
            return { 'visibility': 'hidden' };
        }
        else {
            if (solution.id == this.defaultQuerySolution.id) {
                // //不处理了
            }
            else {
                return { 'visibility': 'hidden' };
            }
        }
    }
    /**
     * 方案选择事件
     * @param {?} solution 当前选中的方案
     * @return {?}
     */
    select(solution) {
        this.selectedSolution = Object.assign(new QuerySolutionVo(), solution);
        if (solution.queryConditions && solution.queryConditions.length > 0) {
            this.selectedFieldConditions = this.selectedSolution.queryConditions;
        }
        else {
            this.selectedFieldConditions = new Array();
        }
        this.showSaveButton = this.selectedSolution.isSystem ? false : true;
        this.visibleSelectedConditions = this.selectedFieldConditions.filter((/**
         * @param {?} item
         * @return {?}
         */
        (item) => { return !!item.visible; }));
        !this.isDisabled && this.solutionHandler.setCurrentQueryConditions(cloneDeep(this.visibleSelectedConditions));
        //this.setCurrentSolutionCache(this.selectedSolution);
        // 如果方案的变更是在収折时
        this.whenFoldSetSummaryConditions();
    }
    /**
     * @param {?} solution
     * @return {?}
     */
    selectAndQuery(solution) {
        //高级模式不使用默认值
        if (solution.mode !== '2' && (this.defaultConfigs && this.defaultConfigs.isCurrent || (solution.isSystem && solution.type !== 'pre'))) {
            this.setQueryConditionValues(this.defaultConfigs, solution); //已包含select操作
        }
        else {
            if (solution.mode !== '2' && this.enableHistory && solution.type == 'pre') {
                /** @type {?} */
                let conditions = cloneDeep(this.historyRefer['foundSchme']['queryConditions']);
                if (conditions.length > 0) {
                    solution.queryConditions = conditions;
                }
            }
            this.select(solution);
        }
        this.query();
    }
    /**
     * 打开字段配置窗口
     * @return {?}
     */
    openFieldConfigDialog() {
        this.farrisQueryCondition.openFieldConfigDialog();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    fieldConditionsChangeHander($event) {
        /** @type {?} */
        let data = $event;
        if (this.selectedSolution.mode === '2') {
            this.selectedSolution.queryConditions = data;
            /** @type {?} */
            let target = this.querySolutionAdvancedVos.find((/**
             * @param {?} val
             * @return {?}
             */
            val => val.id === this.selectedSolution.id));
            if (target) {
                target.queryConditions = this.selectedSolution.queryConditions;
                target.queryConditionsAdvanced = this.selectedSolution.queryConditionsAdvanced;
            }
        }
        // this.selectedSolution.queryConditions = data;此时不改变当前选中的方案，等到保存时，直接调用this.selectedFieldConditions来保存。
        this.selectedFieldConditions = data;
        this.visibleSelectedConditions = this.selectedFieldConditions.filter((/**
         * @param {?} item
         * @return {?}
         */
        (item) => { return !!item.visible; }));
        if (!this.isDisabled) {
            this.solutionHandler.setCurrentQueryConditions(cloneDeep(this.visibleSelectedConditions));
            this.solutionHandler.setFilterConditions(this.visibleSelectedConditions);
            //this.setCurrentSolutionCache(this.selectedSolution);
        }
        if (this.queryAfterValueChange) {
            this.query();
        }
        this.markAsChanged(true, this.selectedSolution.id);
    }
    //高级模式新增，高级模式下不仅需要修改conditions，还需要修改solution的相关属性
    /**
     * @param {?} $event
     * @return {?}
     */
    fieldSolutionChangeHander($event) {
        /** @type {?} */
        let data = $event;
        this.selectedSolution = data;
    }
    /**
     * 保存方案
     * @return {?}
     */
    save() {
        if (!this.formId) {
            this.solutionService.showInfo(this.localeService.getValue("querySolution.container.pleaseInput"), 'error', '', 1500);
            return;
        }
        //id为空则表示没有预置方案，则此时保存方案相当于新增一个方案
        if (!this.selectedSolution.id) {
            this.openSaveAsDialog(this.localeService.getValue("querySolution.container.save"));
        }
        //直接保存
        else {
            /** @type {?} */
            const checkAuthority$ = this.httpService.getAuth('QuerySolution', this.sessionId);
            checkAuthority$.pipe(switchMap((/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                if (!data['success'] && (this.selectedSolution.type == 'public' || this.selectedSolution.type == 'org')) {
                    this.solutionService.showInfo(this.localeService.getValue('extendQuerySolution.auth.updateError'), 'warning', '', 1500);
                    return of(false);
                }
                else {
                    this.selectedSolution.queryConditions = this.selectedFieldConditions;
                    /** @type {?} */
                    const save$ = this.httpService.update(this.selectedSolution, this.sessionId);
                    return save$.pipe(tap((/**
                     * @return {?}
                     */
                    () => {
                        if (this.selectedSolution.mode === '2') {
                            /** @type {?} */
                            let tempIndex = this.querySolutionAdvancedVos.findIndex((/**
                             * @param {?} x
                             * @return {?}
                             */
                            x => x.id == this.selectedSolution.id));
                            // this.selectedFieldConditions = this.selectedSolution.queryConditions;
                            this.querySolutionAdvancedVos.splice(tempIndex, 1, this.selectedSolution);
                        }
                        else {
                            /** @type {?} */
                            let tempIndex = this.querySolutionVos.findIndex((/**
                             * @param {?} x
                             * @return {?}
                             */
                            x => x.id == this.selectedSolution.id));
                            // this.selectedFieldConditions = this.selectedSolution.queryConditions;
                            this.querySolutionVos.splice(tempIndex, 1, this.selectedSolution);
                        }
                        this.markAsChanged(false, this.selectedSolution.id);
                        this.farrisQueryCondition.canClearRelatedFields = false;
                        this.solutionService.showInfo(this.localeService.getValue("querySolution.container.saveSuccess"), 'success', '', 1500);
                    }), (/**
                     * @param {?} error
                     * @return {?}
                     */
                    error => {
                        this.solutionService.showInfo(this.localeService.getValue("querySolution.container.saveFail"), 'error', '', 1500);
                        throw error;
                    })));
                }
            })), tap((/**
             * @return {?}
             */
            () => { }), (/**
             * @param {?} error
             * @return {?}
             */
            error => {
                console.warn(this.localeService.getValue('extendQuerySolution.auth.apiError'));
            }))).subscribe();
        }
    }
    /**
     * 打开另存为对话框，点击保存按钮时也走该逻辑
     * @param {?} title
     * @return {?}
     */
    openSaveAsDialog(title) {
        if (!this.formId) {
            this.solutionService.showInfo(this.localeService.getValue("querySolution.container.pleaseInput"), 'error', '', 1500);
            return;
        }
        this.saveAsDialog.title = title;
        this.saveAsDialog.querySolution = this.newSaveAsSolutionVo();
        this.saveAsDialog.radioGroupValue = 'private';
        this.saveAsDialog.show();
    }
    /**
     * 另存为确定事件处理
     * @param {?} _event
     * @return {?}
     */
    saveAsConfrimHandler(_event) {
        /** @type {?} */
        const solution = _event['querySolution'];
        if (solution.queryConditions && solution.queryConditions.length) {
            solution.queryConditions.forEach((/**
             * @param {?} element
             * @return {?}
             */
            element => {
                if (!!element.readonly) {
                    element.readonly = false;
                }
            }));
        }
        this.httpService.create(solution, this.sessionId).subscribe((/**
         * @return {?}
         */
        () => {
            //重新设置全局的默认方案
            if (solution.isDefault) { //被设置为默认方案
                this.defaultQuerySolution = solution;
                this.querySolutionVos = this.querySolutionVos.map((/**
                 * @param {?} vo
                 * @return {?}
                 */
                vo => {
                    vo.isDefault = false;
                    return vo;
                }));
                this.querySolutionAdvancedVos = this.querySolutionAdvancedVos.map((/**
                 * @param {?} vo
                 * @return {?}
                 */
                vo => {
                    vo.isDefault = false;
                    return vo;
                }));
            }
            //用户保存自己的筛选方案后，不再显示系统默认的方案，操作方案为高级模式时不影响标准
            if (this.querySolutionVos.length === 1 && this.querySolutionVos[0].isSystem && solution.mode === '1') {
                //用户没有配置自己的筛选方案时，或者用户只配置了一个筛选方案时，此数组都有一个方案；若此方案是系统默认方案，则应在另存为时删掉
                this.querySolutionVos = new Array();
            }
            else if (this.querySolutionVos.length === 2 && this.enableHistory) {
                /** @type {?} */
                let filterSolutionVos = new Array();
                filterSolutionVos = this.querySolutionVos.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => {
                    if (item.type == 'pre') {
                        return true;
                    }
                    return !item.isSystem;
                }));
                this.querySolutionVos = filterSolutionVos;
            }
            //高级模式同理
            if (this.querySolutionAdvancedVos.length === 1 && this.querySolutionAdvancedVos[0].isSystem && solution.mode === '2') {
                this.querySolutionAdvancedVos = new Array();
            }
            else if (this.querySolutionAdvancedVos.length === 2 && this.enableHistory) {
                /** @type {?} */
                let filterSolutionVos = new Array();
                filterSolutionVos = this.querySolutionAdvancedVos.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => {
                    if (item.type == 'pre') {
                        return true;
                    }
                    return !item.isSystem;
                }));
                this.querySolutionAdvancedVos = filterSolutionVos;
            }
            solution.mode === '2' ? this.querySolutionAdvancedVos.push(solution) : this.querySolutionVos.push(solution);
            this.selectedSolution = solution;
            this.showSaveButton = this.selectedSolution.isSystem ? false : true;
            this.selectedFieldConditions = this.selectedSolution.queryConditions;
            this.visibleSelectedConditions = this.selectedFieldConditions.filter((/**
             * @param {?} item
             * @return {?}
             */
            (item) => { return !!item.visible; }));
            this.farrisQueryCondition.canClearRelatedFields = false;
            this.saveAsDialog.isDefault = false;
            this.solutionService.showInfo(this.localeService.getValue("querySolution.container.saveSuccess"), 'success', '', 1500);
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            this.solutionService.showInfo(this.localeService.getValue("querySolution.container.saveFail"), 'error', '', 1500);
            throw error;
        }));
    }
    /**
     * 打开管理界面
     * @return {?}
     */
    openManangerDialog() {
        this.httpService.getAuth('QuerySolution', this.sessionId).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            this.manageDialog.authorityError = data['success'] ? false : true;
            this.manageDialog.show(cloneDeep([...this.querySolutionVos, ...this.querySolutionAdvancedVos]));
        }));
    }
    /**
     * 管理界面确定处理
     * @param {?} _event
     * @return {?}
     */
    managerConfirmHandler(_event) {
        /** @type {?} */
        let obj = _event;
        if (obj['validator'] == true) {
            //点击checkbox时，newDefaultQuerySolution被赋值； 全局默认方案被删除时，newDefaultQuerySolution变为null
            /** @type {?} */
            const defaultId = obj['newDefaultQuerySolution'] ? obj['newDefaultQuerySolution'].id : '';
            /** @type {?} */
            const belongId = this.formId;
            //若没有删除任何方案，deletedQuerySolutionVos会是一个空数组
            /** @type {?} */
            const deletedIds = obj['deletedQuerySolutionIds'];
            this.httpService.batchUpdate({ defaultId, belongId, deletedIds }, this.sessionId).subscribe((/**
             * @return {?}
             */
            () => {
                //把删除后的vo赋值给 this.querySolutionVos；如果没删除方案， this.querySolutionVos跟obj['splicedQuerySolutionVos']同值
                //默认方案的变化也记录在updatedQuerySolutionVos中
                this.querySolutionVos = obj['updatedQuerySolutionVos'].filter((/**
                 * @param {?} val
                 * @return {?}
                 */
                val => val.mode !== '2'));
                this.querySolutionAdvancedVos = obj['updatedQuerySolutionVos'].filter((/**
                 * @param {?} val
                 * @return {?}
                 */
                val => val.mode === '2'));
                if (!obj['currentDefaultQuerySolutionId']) { //如果传过来的当前默认方案id是空，代表所有默认方案都取消了
                    this.defaultQuerySolution = null;
                }
                //如果有默认方案先显示默认方案，如果没有默认方案则直接显示当前查询方案的第一个；如果没有任何查询方案则直接清空查询方案，则显示系统预置的默认方案
                if (defaultId) {
                    //已设置默认方案
                    this.select(obj['newDefaultQuerySolution']);
                    this.defaultQuerySolution = obj['newDefaultQuerySolution'];
                    if (this.querySolutionVos.length === 0 || (this.enableHistory && this.querySolutionVos.length === 1 && this.querySolutionVos[0].type == 'pre')) {
                        this.querySolutionVos.push(this.sysQuerySolutionVo);
                    }
                    if (!this.querySolutionAdvancedVos.length) {
                        this.querySolutionAdvancedVos.push(this.initAdvancedSolution());
                    }
                }
                else {
                    //未更改默认方案，判断当前选择的查询方案是否已被删除
                    if (deletedIds.includes(this.selectedSolution.id)) { //如果被删除
                        if (this.querySolutionVos.length > 0) {
                            if (this.enableHistory && this.querySolutionVos.length == 1 && this.querySolutionVos[0].type == 'pre') {
                                this.select(this.sysQuerySolutionVo);
                                this.querySolutionVos.push(this.sysQuerySolutionVo);
                            }
                            else {
                                /** @type {?} */
                                const publicSolution = this.querySolutionVos.find((/**
                                 * @param {?} item
                                 * @return {?}
                                 */
                                item => item.type == 'public'));
                                if (publicSolution) {
                                    this.select(publicSolution);
                                }
                                else {
                                    /** @type {?} */
                                    const orgSolution = this.querySolutionVos.find((/**
                                     * @param {?} item
                                     * @return {?}
                                     */
                                    item => item.type == 'org'));
                                    this.select(!!orgSolution ? orgSolution : cloneDeep(this.querySolutionVos[0]));
                                }
                            }
                        }
                        else {
                            this.select(this.sysQuerySolutionVo);
                            this.querySolutionVos.push(this.sysQuerySolutionVo);
                        }
                        //高级模式没有的话也需要增加默认
                        if (!this.querySolutionAdvancedVos.length) {
                            this.querySolutionAdvancedVos.push(this.initAdvancedSolution());
                        }
                    }
                    else {
                        //什么也没发生
                        //把所有的默认都取消了
                        if (this.querySolutionVos.length === 0 || (this.enableHistory && this.querySolutionVos.length === 1 && this.querySolutionVos[0].type == 'pre')) {
                            this.querySolutionVos.push(this.sysQuerySolutionVo);
                        }
                        if (!this.querySolutionAdvancedVos.length) {
                            this.querySolutionAdvancedVos.push(this.initAdvancedSolution());
                        }
                    }
                }
                this.solutionService.showInfo(this.localeService.getValue("querySolution.container.saveSuccess"), 'success', '', 1500);
            }), (/**
             * @param {?} error
             * @return {?}
             */
            error => {
                this.solutionService.showInfo(this.localeService.getValue("querySolution.container.saveFail"), 'error', '', 1500);
                return Observable.throw(error);
            }));
        }
        else {
            //校验失败
        }
    }
    /**
     * 执行筛选
     * @return {?}
     */
    query() {
        if (!this.isDisabled) {
            /** @type {?} */
            const errorArr = this.solutionService.isValid(this.visibleSelectedConditions);
            //高级模式下不用判断require，有自己的校验逻辑
            if (errorArr && errorArr.length && this.selectedSolution.mode !== '2') {
                /** @type {?} */
                const errorFields = errorArr.map((/**
                 * @param {?} error
                 * @return {?}
                 */
                error => error.field));
                this.solutionService.showInfo(this.localeService.getValue("querySolution.container.require").replace('{fields}', errorFields.join(', ')), 'warning', '', 2500);
            }
            else {
                //高级模式下必须所有条件都有值，否则会影响层级与关联关系
                /** @type {?} */
                let queryConditions = cloneDeep(this.visibleSelectedConditions);
                if (this.selectedSolution.mode === '2') {
                    queryConditions = cloneDeep(this.selectedFieldConditions);
                    /** @type {?} */
                    let emptyFlag = queryConditions.find((/**
                     * @param {?} val
                     * @return {?}
                     */
                    val => val.value.isEmpty()));
                    if (emptyFlag) {
                        this.notify.warning(this.localeService.getValue('advancedMode.auth.conditionEmptyError'));
                        return;
                    }
                }
                this.solutionHandler.setFilterConditions(queryConditions);
                this.queryEmitter.emit(queryConditions);
                // 更新历史记录
                this.changeQueryHistroy();
            }
        }
    }
    /**
     * 清空->条件全部重置为空
     * @return {?}
     */
    clear() {
        // this.isFold = !this.isFold;
        /** @type {?} */
        const isNotEmpty = this.visibleSelectedConditions.some((/**
         * @param {?} condition
         * @return {?}
         */
        condition => condition.value.isEmpty() == false));
        if (isNotEmpty) {
            /** @type {?} */
            const newList = cloneDeep(this.farrisQueryCondition.resetSelectedConditionValues(true));
            this.selectedFieldConditions = newList;
            this.selectedSolution.queryConditions = newList;
            this.visibleSelectedConditions = this.selectedFieldConditions.filter((/**
             * @param {?} item
             * @return {?}
             */
            (item) => { return !!item.visible; }));
            this.solutionHandler.setCurrentQueryConditions(cloneDeep(this.visibleSelectedConditions));
            this.solutionHandler.setFilterConditions(this.visibleSelectedConditions);
            this.markAsChanged(true, this.selectedSolution.id);
            this.query();
        }
    }
    /**
     * 折叠按钮的清空
     * @return {?}
     */
    foldClearHandler() {
        if (this.isFold) {
            // 如果点击的时候是収折的
            this.isFold = false;
            if (this.summaryConditions.length > 0) {
                this.summaryConditions = [];
                this.clear();
            }
        }
        else {
            // 如果点击的时候是展开的
            this.clear();
        }
    }
    /**
     * 箭头点击事件
     * @return {?}
     */
    arrowClick() {
        this.isFold = !this.isFold;
        this.whenFoldSetSummaryConditions();
        if (!this.isFold) {
            this.summaryConditions = [];
        }
    }
    /**
     * @private
     * @return {?}
     */
    whenFoldSetSummaryConditions() {
        if (this.isFold) {
            // 如果収折
            /** @type {?} */
            const isNotEmpty = this.visibleSelectedConditions.some((/**
             * @param {?} condition
             * @return {?}
             */
            condition => condition.value.isEmpty() == false));
            if (isNotEmpty) {
                this.summaryConditions = this.visibleSelectedConditions.filter((/**
                 * @param {?} condition
                 * @return {?}
                 */
                condition => condition.value.isEmpty() == false));
            }
            else {
                this.summaryConditions = [];
            }
        }
    }
    /**
     * 为系统预置方案或者自定义查询方案设置筛选条件默认值
     * @param {?} configs 默认值设置
     * @param {?} selectedSolution 当前选择的筛选方案
     * @return {?}
     */
    setQueryConditionValues(configs, selectedSolution) {
        /** @type {?} */
        let targetSolution;
        /** @type {?} */
        let targetSolutionWithValues;
        //高级模式下不支持默认值
        if (configs && Array.isArray(configs.fieldConfigs) && configs.fieldConfigs.length && selectedSolution.mode !== '2') {
            if (selectedSolution.hasOwnProperty('type') && selectedSolution['type'] == 'pre') {
                //如果是历史记录
                targetSolution = null;
                this.select(selectedSolution);
            }
            else {
                targetSolution = configs.isCurrent ? selectedSolution : (selectedSolution.isSystem ? selectedSolution : null);
            }
            if (targetSolution && targetSolution.queryConditions && targetSolution.queryConditions.length) {
                targetSolutionWithValues = cloneDeep(this.solutionService.setValues(targetSolution, configs.fieldConfigs, !!configs.readonly));
                this.doneSetDefaultValue = true;
                this.select(targetSolutionWithValues);
            }
            else {
                //不为自定义方案设置默认值时，标记下要为其赋默认值的意图，但未成功
                this.doneSetDefaultValue = true;
                //如果没有设置默认值，就把当前选中方案的条件的值传出
                this.solutionHandler.setCurrentQueryConditions(this.selectedSolution.queryConditions);
            }
        }
        else {
            this.select(selectedSolution);
        }
    }
    /**
     * 加载筛选方案后自动查询
     * @param {?} currentSolution 当前筛选方案
     * @return {?}
     */
    initAutoQuery(currentSolution) {
        if (this.enableInitQuery && this.doneInitQuery && this.enableHistory && currentSolution.mode !== '2') {
            // 已通过历史记录进行过查询
            return;
        }
        /** @type {?} */
        let enableQuery = true;
        if (currentSolution && currentSolution.queryConditions && currentSolution.queryConditions.length) {
            if (this.isSetDefaultValue && !this.doneSetDefaultValue) {
                enableQuery = false;
            }
        }
        else {
            enableQuery = false;
        }
        if (enableQuery && this.enableInitQuery) {
            // this.solutionHandler.setFilterConditions(this.selectedSolution.queryConditions); 执行query方法时，已经执行setFilterConditions
            //if(this._self.nativeElement.querySelector('.farris-panel .btn-group .btn-primary')){
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.query();
                this.doneInitQuery = true;
            }));
            //}
        }
    }
    /**
     * 当前查询方案的筛选条件发生变化时，标记方案处于编辑状态
     * @param {?} isChanged 方案是否处于编辑状态
     * @param {?} solutionId 需要标记是否处于编辑状态的方案的id
     * @return {?}
     */
    markAsChanged(isChanged, solutionId) {
        this.querySolutionVos.forEach((/**
         * @param {?} solution
         * @return {?}
         */
        solution => {
            if (solution.id == solutionId)
                solution.isChanged = isChanged;
        }));
        this.querySolutionAdvancedVos.forEach((/**
         * @param {?} solution
         * @return {?}
         */
        solution => {
            if (solution.id == solutionId)
                solution.isChanged = isChanged;
        }));
        this.selectedSolution.isChanged = isChanged;
        this.cdf.detectChanges();
    }
    /**
         * 处理筛选按钮和清空按钮的响应式位置变更
         * @param $event farris-querycondition组件内部是否显示按钮
         */
    // showButtonHandle($event: any) {
    //     if(!this.isDisabled){
    //         let data = $event;
    //         this.queryButtonOnRight = !data;
    //         if(this.enableInitQuery && !this.doneInitQuery) {
    //             this.initAutoQuery(this.selectedSolution);
    //         }
    //     }
    // }
    /**
     * 键盘监听函数，当按下回车键时相当于按下筛选按钮
     * @private
     * @return {?}
     */
    listenKeyBoard() {
        this._self.nativeElement.addEventListener("keydown", this.handleKeyDown);
    }
    /**
     * 自动初始化查询
     * 启用历史记录，并且启用了初始化查询，如果有条件，就进行查询
     * @private
     * @return {?}
     */
    initAutoQueryByLS() {
        // 启用历史
        if (this.enableHistory && !this.isDisabled) {
            // key
            this.setLSKey();
            // 初始化对象
            this.initQuerySolution();
            // 如果设置了初始查询
            if (this.enableInitQuery) {
                this.formatInitQuerySolution();
                if (this.historyRefer['foundSchme']['queryConditions'].length > 0) {
                    /** @type {?} */
                    const queryConditions = cloneDeep(this.historyRefer['foundSchme']['queryConditions']);
                    this.selectedFieldConditions = queryConditions;
                    this.visibleSelectedConditions = this.selectedFieldConditions.filter((/**
                     * @param {?} item
                     * @return {?}
                     */
                    (item) => { return !!item.visible; }));
                    /** @type {?} */
                    const visibleConditions = cloneDeep(this.visibleSelectedConditions);
                    // 设置查询条件
                    this.solutionHandler.setFilterConditions(visibleConditions);
                    // 抛出查询事件
                    this.queryEmitter.emit(visibleConditions);
                    // 标记
                    this.doneInitQuery = true;
                }
            }
        }
    }
    /**
     * 初始化
     * @private
     * @return {?}
     */
    initQuerySolution() {
        /** @type {?} */
        let querysolutionVO = new QuerySolutionVo();
        querysolutionVO.id = this.solutionService.getGuid();
        querysolutionVO.code = this.localeService.getValue("querySolution.container.histroyName");
        querysolutionVO.queryConditions = [];
        querysolutionVO.belongId = this.formId;
        querysolutionVO.isSystem = true;
        querysolutionVO.type = 'pre';
        querysolutionVO.mode = '1';
        querysolutionVO.queryConditionsAdvanced = { relation: 1, items: [], children: [], path: [1], groupid: 1 };
        this.historyRefer['foundSchme'] = querysolutionVO;
    }
    /**
     * 初始化高级模式
     * @private
     * @return {?}
     */
    initAdvancedSolution() {
        /** @type {?} */
        let querysolutionVO = new QuerySolutionVo();
        querysolutionVO.id = this.solutionService.getGuid();
        querysolutionVO.code = ((this.presetQuerySolutionName != this.localeService.getValue("querySolution.container.sysPresetName")) ? this.presetQuerySolutionName : this.localeService.getValue("querySolution.container.defaultName")) + this.localeService.getValue("advancedMode.text.sysAdvancedName");
        ;
        querysolutionVO.queryConditions = [];
        querysolutionVO.belongId = this.formId;
        querysolutionVO.isSystem = true;
        querysolutionVO.type = 'pre';
        querysolutionVO.mode = '2';
        querysolutionVO.queryConditionsAdvanced = { relation: 1, items: [], children: [], path: [1], groupid: 1 };
        return querysolutionVO;
    }
    /**
     * 初始化来自LS的条件
     * @private
     * @return {?}
     */
    formatInitQuerySolution() {
        /** @type {?} */
        let histroyCondition = this.httpService.getLS(this.historyRefer.schemeKeyLS);
        if (histroyCondition.length > 0) {
            /** @type {?} */
            let convertor = new QuerySolutionVoConvertor();
            this.historyRefer['foundSchme']['queryConditions'] = histroyCondition;
            // 规范化条件
            this.historyRefer['foundSchme'] = this.formatQuerySolutionVO(convertor, this.historyRefer['foundSchme']);
            // 更新配置
            this.solutionService.updateControl(this.historyRefer['foundSchme'], this.resetFieldConfigs);
            this.historyRefer.initCondition = true;
        }
    }
    /**
     * 保存之后更新历史
     * @private
     * @param {?=} updateAll
     * @return {?}
     */
    changeQueryHistroy(updateAll = true) {
        if (!this.enableHistory || this.selectedSolution.mode !== '1') {
            return;
        }
        /** @type {?} */
        let preIndex = this.querySolutionVos.findIndex((/**
         * @param {?} qsolutionVO
         * @return {?}
         */
        (qsolutionVO) => {
            return qsolutionVO['type'] == 'pre';
        }));
        // 执行查询之后，上次查询的修改就不认为是变更了
        if (preIndex > -1) {
            this.querySolutionVos[preIndex]['isChanged'] = false;
        }
        if (this.selectedSolution['type'] == 'pre') {
            this.selectedSolution['isChanged'] = false;
        }
        /** @type {?} */
        const queryConditions = cloneDeep(this.selectedFieldConditions);
        this.historyRefer['foundSchme'].queryConditions = queryConditions;
        // 更新缓存
        this.httpService.setLS(this.historyRefer.schemeKeyLS, this.historyRefer['foundSchme'].queryConditions);
        if (updateAll) {
            // 发起请求保存历史记录
            this.httpService.saveHistroy(this.historyRefer['foundSchme'], this.sessionId).subscribe((/**
             * @return {?}
             */
            () => {
            }));
        }
    }
    /**
     * 处理iframe嵌套
     * @private
     * @return {?}
     */
    getTopWindow() {
        /** @type {?} */
        let curWin = window.self;
        try {
            while (curWin.location.origin === curWin.parent.location.origin) {
                if (curWin !== curWin.parent && !((/** @type {?} */ (curWin))).isRTFTopWin) {
                    curWin = curWin.parent;
                }
                else {
                    break;
                }
            }
        }
        catch (e) { }
        return curWin;
    }
    /**
     * 设置key
     * @private
     * @return {?}
     */
    setLSKey() {
        // 获取window
        /** @type {?} */
        let curWindow = this.getTopWindow();
        /** @type {?} */
        let userCode = '';
        if (curWindow['gspframeworkService'] && curWindow['gspframeworkService'].hasOwnProperty('common')) {
            userCode = curWindow['gspframeworkService']['common'].userInfos.get().code;
        }
        this.historyRefer.schemeKeyLS = userCode ? 'Scheme' + this._idSer.encrypt(userCode + this.formId) : '';
    }
    // 判断当前用户是否有权限设置公共类型筛选方案
    /**
     * @param {?} e
     * @return {?}
     */
    checkAuthority(e) {
        if (e != 'private') {
            this.httpService.getAuth('QuerySolution', this.sessionId).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            data => {
                if (data) {
                    if (!data['success']) {
                        this.solutionService.showInfo(this.localeService.getValue('extendQuerySolution.auth.createError'), 'warning', '', 1500);
                        this.saveAsDialog.authorityError = true;
                        this.saveAsDialog.radioGroupValue = this.saveAsDialog.pRadioGroupValue;
                    }
                    else {
                        if (e == 'org') {
                            /** @type {?} */
                            const orgData = data['authOrganizationData'];
                            if (Array.isArray(orgData) && orgData.length) {
                                this.saveAsDialog.dataPermissionIds = orgData;
                                if (orgData.toString() != '*' && !this.saveAsDialog.selectedOrgIds.length) {
                                    this.saveAsDialog.selectedOrgIds = this.saveAsDialog.dataPermissionIds;
                                }
                                this.saveAsDialog.noDataAuthority = false;
                                this.saveAsDialog.pRadioGroupValue = e;
                            }
                            else {
                                this.solutionService.showInfo(this.localeService.getValue('extendQuerySolution.auth.noDataAuthority'), 'warning', '', 1500);
                                this.saveAsDialog.noDataAuthority = true;
                                this.saveAsDialog.radioGroupValue = this.saveAsDialog.pRadioGroupValue;
                            }
                        }
                        else {
                            this.saveAsDialog.pRadioGroupValue = e;
                        }
                    }
                }
                else {
                    console.warn(this.localeService.getValue('extendQuerySolution.auth.apiError'));
                }
            }));
        }
        else {
            this.saveAsDialog.pRadioGroupValue = e;
            this.saveAsDialog.authorityError = false;
        }
    }
}
QuerySolutionComponent.decorators = [
    { type: Component, args: [{
                selector: 'farris-querysolution',
                template: "<div #querysolution class=\"farris-panel position-relative\" tabindex=\"1\" style=\"outline: none;\">\r\n    <div class=\"solution-header\">\r\n        <div class=\"btn-group mr-3\" fDropdown>\r\n            <div class=\"solution-header-title\" fDropdownToggle>               \r\n                <span style=\"max-width: 288px;overflow: hidden;text-overflow: ellipsis;white-space: nowrap\"\r\n                    title=\"{{selectedSolution.code}}\">{{selectedSolution.code}}</span>\r\n                <b class=\"header-title--change\" *ngIf=\"selectedSolution.isChanged\">*</b>\r\n                <span class=\"f-icon f-accordion-expand\"></span>\r\n            </div>\r\n            <div class=\"query-solution dropdown-menu solution-header-title-menu\" fDropdownMenu>\r\n                <div class=\"solution-header-title-menu-inner\">\r\n                    <div class=\"solution-header-title-menu-arrow\"></div>\r\n                    <span class=\"solution-header-managerList-type ml-2 mb-2\" *ngIf=\"openAdvancedMode\">{{'advancedMode.text.standardMode' | locale}}</span>\r\n                    <li class=\"dropdown-item solution-header-dropdown-item\"\r\n                        [ngClass]=\"{'solution-header-dropdown-item-active': solution.id === selectedSolution.id }\"\r\n                        (click)=\"selectAndQuery(solution)\" *ngFor=\"let solution of querySolutionVos\">\r\n                        <b *ngIf=\"solution.isChanged\"  class=\"menu-item--change\" >*</b>\r\n                        <span class=\"solution-header-dropdown-item-title\" [title]=\"solution.code\">{{solution.code}}</span>\r\n                        <span class=\"solution-header-dropdown-item-tip\"\r\n                            [ngStyle]=\"getDefaultLabel(solution)\">{{'querySolution.container.default' | locale}}</span>\r\n                        <span class=\"solution-header-dropdown-item-public\" *ngIf=\"solution.type == 'public'\">\r\n                            {{'extendQuerySolution.public.label' | locale}}</span>\r\n                        <span class=\"solution-header-dropdown-item-public\" *ngIf=\"solution.type == 'org'\">\r\n                            {{'extendQuerySolution.org.label' | locale}}</span>\r\n                    </li>\r\n                    <!-- <li class=\"dropdown-item solution-header-dropdown-item-btns\" style=\"background-color: white!important;cursor: default;\">\r\n                        <span class=\"dropdown-item-btn\" *ngIf=\"showSaveButton\"\r\n                            (click)=\"save()\">{{'queryCondition.container.save' | locale}}</span>\r\n                        <span class=\"dropdown-item-btn\"\r\n                            (click)=\"saveAs()\">{{'queryCondition.container.saveAs' | locale}}</span>\r\n                        <span class=\"dropdown-item-btn\"\r\n                            (click)=\"openManangerDialog()\">{{'querySolution.container.manage' | locale}}</span>\r\n                    </li> -->\r\n                    <ng-container *ngIf=\"openAdvancedMode\">\r\n                        <span class=\"solution-header-managerList-type ml-2 mb-2\">{{'advancedMode.text.advancedMode' | locale}}</span>\r\n                        <li class=\"dropdown-item solution-header-dropdown-item\"\r\n                            [ngClass]=\"{'solution-header-dropdown-item-active': solution.id === selectedSolution.id }\"\r\n                            (click)=\"selectAndQuery(solution)\" *ngFor=\"let solution of querySolutionAdvancedVos\">\r\n                            <b *ngIf=\"solution.isChanged\" class=\"menu-item--change\">*</b>\r\n                            <span class=\"solution-header-dropdown-item-title\" [title]=\"solution.code\">{{solution.code}}</span>\r\n                            <span class=\"solution-header-dropdown-item-tip\"\r\n                                [ngStyle]=\"getDefaultLabel(solution)\">{{'querySolution.container.default' | locale}}</span>\r\n                            <span class=\"solution-header-dropdown-item-public\" *ngIf=\"solution.type == 'public'\">\r\n                                {{'extendQuerySolution.public.label' | locale}}</span>\r\n                            <span class=\"solution-header-dropdown-item-public\" *ngIf=\"solution.type == 'org'\">\r\n                                {{'extendQuerySolution.org.label' | locale}}</span>\r\n                        </li>\r\n                    </ng-container>\r\n                    <li class=\"dropdown-item solution-header-dropdown-item-btns\" style=\"background-color: white!important;cursor: default;\">\r\n                        <span class=\"dropdown-item-btn\" *ngIf=\"showSaveButton\"\r\n                            (click)=\"save()\">{{'queryCondition.container.save' | locale}}</span>\r\n                        <span class=\"dropdown-item-btn\"\r\n                            (click)=\"saveAs()\">{{'queryCondition.container.saveAs' | locale}}</span>\r\n                        <span class=\"dropdown-item-btn\"\r\n                            (click)=\"openManangerDialog()\">{{'querySolution.container.manage' | locale}}</span>\r\n                    </li>\r\n                </div>\r\n            </div>\r\n        </div>\r\n        <div class=\"summary-condition\" *ngIf=\"isFold && summaryConditions.length > 0\">\r\n            <lib-summary-list [conditions]=\"summaryConditions\"></lib-summary-list>\r\n        </div>\r\n        <div class=\"solution-action\">\r\n            <div class=\"btn-group\" *ngIf=\"!isFold\">\r\n                <button *ngIf=\"isFold==false\" (click)=\"query()\" type=\"button\"\r\n                    class=\"btn btn-primary \">{{filterText}}</button>\r\n            </div>\r\n            <div class=\"icon-group\">\r\n                <!--\u6E05\u7A7A\u6309\u94AE-->\r\n                <span *ngIf=\"(!isFold || summaryConditions.length > 0) && selectedSolution.mode !== '2'\" class=\"icon-group-remove\" (click)=\"foldClearHandler()\"\r\n                    title=\"{{'querySolution.container.clear' | locale}}\">\r\n                    <span class=\"f-icon f-icon-remove\"></span>\r\n                </span>                \r\n                <ng-container *ngIf=\"!isFold\">\r\n                    <span class=\"divide\"></span>\r\n                    <!--\u914D\u7F6E\u6309\u94AE-->\r\n                    <span *ngIf=\"selectedSolution.mode !== '2'\" class=\"icon-group-setup\" (click)=\"openFieldConfigDialog()\"\r\n                        title=\"{{'queryCondition.container.config' | locale}}\">\r\n                        <span class=\"f-icon f-icon-home-setup\"></span>\r\n                    </span>\r\n                    <!--\u5C55\u5F00\u6536\u8D77-->\r\n                    <span class=\"icon-group-packup\" (click)=\"arrowClick()\"\r\n                        title=\"{{'querySolution.container.arrowUp' | locale}}\">\r\n                        <span class=\"f-icon f-icon-packup\"></span>\r\n                    </span>\r\n                </ng-container>\r\n                <!--\u5C55\u5F00\u6536\u8D77-->\r\n                <span *ngIf=\"isFold\" class=\"icon-group-packup\" (click)=\"arrowClick()\"\r\n                    title=\"{{'querySolution.container.arrowDown' | locale}}\">\r\n                    <span class=\"f-icon f-icon-packup\" style=\"transform: rotate(180deg);\"></span>\r\n                </span>\r\n\r\n            </div>\r\n        </div>\r\n    </div>\r\n    <div>\r\n        <farris-querycondition [ngClass]=\"{'d-none':isFold}\" #farrisQueryCondition\r\n            [(selectedFieldConditions)]=\"selectedFieldConditions\" [fieldConfigs]=\"fieldConfigs\"\r\n            [isControlInline]=\"isControlInline\" [showCompleteLabel]=\"showCompleteLabel\"\r\n            [isDisabled]=\"isDisabled\" [resetFieldConfigs]=\"resetFieldConfigs\"\r\n            [unformatFieldConfigs]=\"unformatFieldConfigs\"\r\n            (fieldConditionsChange)=\"fieldConditionsChangeHander($event)\"\r\n            [selectSolution]=\"selectedSolution\">\r\n        </farris-querycondition>\r\n    </div>\r\n\r\n    <app-solutionmanager #manageDialog (confirmEmitter)=\"managerConfirmHandler($event)\"></app-solutionmanager>\r\n    <app-saveasdialog #saveAsDialog\r\n        (confirmEmitter)=\"saveAsConfrimHandler($event)\"\r\n        (getAuthority)=\"checkAuthority($event)\"\r\n        [hideOrgselector]=\"hideOrgselector\"></app-saveasdialog>\r\n    <div *ngIf=\"isDisabled\" style=\"position:absolute;top:0;bottom:0;left:0;right:0;z-index: 99;\"></div>\r\n</div>",
                encapsulation: ViewEncapsulation.None,
                providers: [HttpService, ConvertorService, FieldConfigService],
                styles: [".solution-action{float:right}.solution-action span{vertical-align:middle;margin:0 3px}.solution-action .divide{font-size:12px;padding:3px 0}.solution-header{height:30px;margin-bottom:5px;-webkit-box-pack:start;justify-content:flex-start}.solution-action .f-icon{cursor:pointer}.solution-action .icon-group{display:inline-block}.solution-header-title-text{overflow:initial!important}.query-solution .solution-header-dropdown-item .solution-header-dropdown-item-public{flex-shrink:0;height:1.125rem;min-width:2.5rem;padding:0 4px;line-height:1.125rem;font-size:.75rem;background:#2a87ff;color:#fff;border-radius:2px;text-align:center}.solution-header-managerList-type{font-weight:700}"]
            }] }
];
/** @nocollapse */
QuerySolutionComponent.ctorParameters = () => [
    { type: NotifyService },
    { type: ComponentFactoryResolver },
    { type: Injector },
    { type: BsModalService },
    { type: ConvertorService },
    { type: SolutionService },
    { type: HttpService },
    { type: FieldConfigService },
    { type: LocaleService }
];
QuerySolutionComponent.propDecorators = {
    defaultConfigs: [{ type: Input }],
    fieldConfigs: [{ type: Input }],
    presetQuerySolutionName: [{ type: Input }],
    resetFieldConfigs: [{ type: Input }],
    presetFieldConfigs: [{ type: Input }],
    isDisabled: [{ type: Input }],
    formId: [{ type: Input }],
    queryEmitter: [{ type: Output }],
    cls: [{ type: HostBinding, args: ['class',] }],
    isFold: [{ type: Input }],
    isControlInline: [{ type: Input }],
    showCompleteLabel: [{ type: Input }],
    enableInitQuery: [{ type: Input }],
    enableHistory: [{ type: Input }],
    openAdvancedMode: [{ type: Input }],
    filterText: [{ type: Input }],
    queryAfterValueChange: [{ type: Input }],
    hideOrgselector: [{ type: Input }],
    farrisQueryCondition: [{ type: ViewChild, args: ['farrisQueryCondition',] }],
    manageDialog: [{ type: ViewChild, args: ['manageDialog',] }],
    saveAsDialog: [{ type: ViewChild, args: ['saveAsDialog',] }]
};
if (false) {
    /** @type {?} */
    QuerySolutionComponent.prototype._fieldConfigs;
    /** @type {?} */
    QuerySolutionComponent.prototype._presetFieldConfigs;
    /** @type {?} */
    QuerySolutionComponent.prototype._resetFieldConfigs;
    /**
     *  筛选方案默认值配置
     * @type {?}
     */
    QuerySolutionComponent.prototype.defaultConfigs;
    /**
     * 高级模式需要原始fieldConfigs，需要将范围类控件改为单值控件
     * @type {?}
     */
    QuerySolutionComponent.prototype.unformatFieldConfigs;
    /**
     * 系统预置查询方案名称
     * @type {?}
     */
    QuerySolutionComponent.prototype.presetQuerySolutionName;
    /**
     * 此值由设计器传入
     * 组件在设计时无法点击被禁用，true
     * 组件在运行时可以点击使用，false
     * @type {?}
     */
    QuerySolutionComponent.prototype.isDisabled;
    /**
     * 表单ID，必选字段，表示查询方案所属的表单
     * @type {?}
     */
    QuerySolutionComponent.prototype.formId;
    /**
     * sessionId
     * @type {?}
     */
    QuerySolutionComponent.prototype.sessionId;
    /**
     * 查询条件列表
     * @type {?}
     */
    QuerySolutionComponent.prototype.queryEmitter;
    /**
     * 查询方案cls
     * @type {?}
     */
    QuerySolutionComponent.prototype.cls;
    /**
     * 查询方案数据源
     * @type {?}
     */
    QuerySolutionComponent.prototype.querySolutionVos;
    /** @type {?} */
    QuerySolutionComponent.prototype.querySolutionAdvancedVos;
    /**
     * 当前所选的方案，有可能不是方案
     * @type {?}
     */
    QuerySolutionComponent.prototype.selectedSolution;
    /**
     * 是否折叠
     * @type {?}
     */
    QuerySolutionComponent.prototype.isFold;
    /**
     * label和控件是否排成一行，默认是；false时排成上下两行
     * @type {?}
     */
    QuerySolutionComponent.prototype.isControlInline;
    /**
     * 是否需要完整显示超过6个字符的标签，默认false
     * @type {?}
     */
    QuerySolutionComponent.prototype.showCompleteLabel;
    /**
     * 是否启用加载后查询
     * @type {?}
     */
    QuerySolutionComponent.prototype.enableInitQuery;
    /**
     * 默认方案
     * @type {?}
     */
    QuerySolutionComponent.prototype.defaultQuerySolution;
    /**
     * 选择的控件展示列表
     * @type {?}
     */
    QuerySolutionComponent.prototype.selectedFieldConditions;
    /**
     * visible属性为true的过滤条件，是变量selectedFieldConditions的子级
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.visibleSelectedConditions;
    /** @type {?} */
    QuerySolutionComponent.prototype.sysQuerySolutionVo;
    /** @type {?} */
    QuerySolutionComponent.prototype.showSaveButton;
    /** @type {?} */
    QuerySolutionComponent.prototype.isSetDefaultValue;
    /** @type {?} */
    QuerySolutionComponent.prototype.doneSetDefaultValue;
    /** @type {?} */
    QuerySolutionComponent.prototype.doneInitQuery;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.conditionVisibleState;
    /** @type {?} */
    QuerySolutionComponent.prototype.destroy$;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.templateSimpleConditions;
    /**
     * 组件服务接口
     * @type {?}
     */
    QuerySolutionComponent.prototype.solutionHandler;
    /** @type {?} */
    QuerySolutionComponent.prototype.cdf;
    /**
     * 条件概要
     * @type {?}
     */
    QuerySolutionComponent.prototype.summaryConditions;
    /** @type {?} */
    QuerySolutionComponent.prototype.eventManager;
    /**
     * 是否启用记录历史方式
     * 上次查询——不能删除、可以另存、不存在变更
     * 方案列表只有上次查询时，上次查询、系统默认都存在
     * 方案列表多个时，有上次查询，上次查询选中
     * 方案列表没有上次查询时，但是LocalStorage有,构造一个上次查询
     *
     * @type {?}
     */
    QuerySolutionComponent.prototype.enableHistory;
    /**
     * 是否启动高级模式
     * @type {?}
     */
    QuerySolutionComponent.prototype.openAdvancedMode;
    /** @type {?} */
    QuerySolutionComponent.prototype.filterText;
    /** @type {?} */
    QuerySolutionComponent.prototype.queryAfterValueChange;
    /** @type {?} */
    QuerySolutionComponent.prototype.hideOrgselector;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.historyRefer;
    /** @type {?} */
    QuerySolutionComponent.prototype.farrisQueryCondition;
    /** @type {?} */
    QuerySolutionComponent.prototype.manageDialog;
    /** @type {?} */
    QuerySolutionComponent.prototype.saveAsDialog;
    /** @type {?} */
    QuerySolutionComponent.prototype._self;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype._idSer;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.msgSer;
    /** @type {?} */
    QuerySolutionComponent.prototype.newSaveAsSolutionVo;
    /** @type {?} */
    QuerySolutionComponent.prototype.handleKeyDown;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.convertService;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.solutionService;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.httpService;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.fieldConfigService;
    /**
     * @type {?}
     * @private
     */
    QuerySolutionComponent.prototype.localeService;
}
