// @gsp-svc/expression/esm5/lib/expression-var-info-grid/expression-var-info-grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input, ViewChild } from '@angular/core';
import { VarType } from '../expression-var-grid/vartype';
import { ExpressionEventService } from '../expression-event-service';
import { DatagridComponent } from '@farris/ui-datagrid';
import { ExpressionEditorLocalePipe } from '../expression-editor-locale.pipe';
var ExpressionVarInfoGridComponent = /** @class */ (function () {
    // private http = new HttpClient();
    function ExpressionVarInfoGridComponent(eventService) {
        this.eventService = eventService;
        this.currentVarType = VarType.Context;
        this.contextEntities = [];
        this.sessionEntities = [];
        this.selectVarInfoRow = new EventEmitter();
        this.dbclickInsert = new EventEmitter();
        this.items = [];
        this.columns = [];
        //初始化i18n标识
        this.localeId = localStorage.getItem('languageCode');
        this.localePipe = new ExpressionEditorLocalePipe(this.localeId);
        this.columns = [
            { field: 'key', title: this.localePipe.transform("VariableName"), width: '20%', filter: true },
            { field: 'name', title: this.localePipe.transform("VariableSemanticNameValue"), width: '30%', filter: true }
        ];
    }
    /**
     * @return {?}
     */
    ExpressionVarInfoGridComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.eventService.subClearFuncSelect().subscribe((/**
         * @return {?}
         */
        function () {
            _this.dg.clearAll();
        }));
    };
    /**
     * @return {?}
     */
    ExpressionVarInfoGridComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        switch (this.currentVarType) {
            case VarType.Context:
                this.items = this.contextEntities;
                break;
            case VarType.Session:
                this.items = this.sessionEntities;
                break;
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    ExpressionVarInfoGridComponent.prototype.ondblclickRow = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.dbclickInsert.emit($event.rowData);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    ExpressionVarInfoGridComponent.prototype.selectRowChange = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.selectVarInfoRow.emit($event.data);
    };
    ExpressionVarInfoGridComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-expression-var-info-grid',
                    template: "<farris-datagrid #dg\r\n[idField]=\"'key'\"\r\n[height]=\"268\"\r\n[width]=\"430\"\r\n[columns]=\"columns\" \r\n[data]=\"items\"\r\n[fit]=\"true\"\r\n[fitColumns]=\"true\"\r\n[striped]=\"false\"\r\n[showHeader]=\"true\"\r\n[enableFilterRow]=\"true\"\r\n[pagination]=\"false\"\r\n(dblclickRow)=\"ondblclickRow($event)\"\r\n(selectChanged)=\"selectRowChange($event)\"\r\n>\r\n</farris-datagrid>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ExpressionVarInfoGridComponent.ctorParameters = function () { return [
        { type: ExpressionEventService }
    ]; };
    ExpressionVarInfoGridComponent.propDecorators = {
        dg: [{ type: ViewChild, args: ['dg',] }],
        currentVarType: [{ type: Input }],
        contextEntities: [{ type: Input }],
        sessionEntities: [{ type: Input }],
        selectVarInfoRow: [{ type: Output }],
        dbclickInsert: [{ type: Output }]
    };
    return ExpressionVarInfoGridComponent;
}());
export { ExpressionVarInfoGridComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ExpressionVarInfoGridComponent.prototype.localeId;
    /**
     * @type {?}
     * @private
     */
    ExpressionVarInfoGridComponent.prototype.localePipe;
    /** @type {?} */
    ExpressionVarInfoGridComponent.prototype.dg;
    /** @type {?} */
    ExpressionVarInfoGridComponent.prototype.currentVarType;
    /** @type {?} */
    ExpressionVarInfoGridComponent.prototype.contextEntities;
    /** @type {?} */
    ExpressionVarInfoGridComponent.prototype.sessionEntities;
    /** @type {?} */
    ExpressionVarInfoGridComponent.prototype.selectVarInfoRow;
    /** @type {?} */
    ExpressionVarInfoGridComponent.prototype.dbclickInsert;
    /** @type {?} */
    ExpressionVarInfoGridComponent.prototype.items;
    /** @type {?} */
    ExpressionVarInfoGridComponent.prototype.columns;
    /**
     * @type {?}
     * @private
     */
    ExpressionVarInfoGridComponent.prototype.eventService;
}
