// @farris/ui-batch-edit-dialog/esm5/lib/batch-edit-dialog.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ComponentFactoryResolver, Injector, EventEmitter } from '@angular/core';
import { BatchEditDialogComponent } from './batch-edit-dialog.component';
import { BsModalService } from '@farris/ui-modal';
import { LocaleService } from '@farris/ui-locale';
import { of, Subject } from 'rxjs';
import { ResultConfirmComponent } from './result-confirm/result-confirm.component';
// tslint:disable: max-line-length
var BatchEditDialogService = /** @class */ (function () {
    function BatchEditDialogService(injector, modalService, componentFactoryResolver, localeService) {
        this.injector = injector;
        this.modalService = modalService;
        this.componentFactoryResolver = componentFactoryResolver;
        this.localeService = localeService;
        this.dialogRef = null;
        this.componentRef = null;
        // 用户对输入进行确认，此时可以更新bindingData数据
        this.onUserConfirmed = new EventEmitter();
    }
    /**
     * @description 控件参数格式
     * # 控件参数格式
     * ## textbox
     * ```json
     * {"title":"姓名","field":"name","editor":{"controlType":EditorTypes.TEXTBOX}}
     * ```
     * ## dropdown
     * ```json
     * {"title":"姓名","field":"name","editor":{"controlType":EditorTypes.COMBOLIST,"idField":"id","textField":"text","items":[],"multiSelect":false}}
     * ```
     * ## lookup
     * ```json
     * {"title":"姓名","field":"name","editor":{"controlType":EditorTypes.LOOKUP,"uri":"","idField":"id","textField":"text","valueField":"id","displayType":"LIST","mapFields":{},"singleSelect":true,"dialogTitle":"","hiddenInputName":"","items":[],"multiSelect":false}}
     * ```
     * ## 数值输入框
     * ```json
     * {"title":"数值","field":"name","editor":{"controlType":EditorTypes.NUMBER,"precision":2,"step":1,"min":1,"max":10}}
     * ```
     */
    /**
     * \@description 控件参数格式
     * # 控件参数格式
     * ## textbox
     * ```json
     * {"title":"姓名","field":"name","editor":{"controlType":EditorTypes.TEXTBOX}}
     * ```
     * ## dropdown
     * ```json
     * {"title":"姓名","field":"name","editor":{"controlType":EditorTypes.COMBOLIST,"idField":"id","textField":"text","items":[],"multiSelect":false}}
     * ```
     * ## lookup
     * ```json
     * {"title":"姓名","field":"name","editor":{"controlType":EditorTypes.LOOKUP,"uri":"","idField":"id","textField":"text","valueField":"id","displayType":"LIST","mapFields":{},"singleSelect":true,"dialogTitle":"","hiddenInputName":"","items":[],"multiSelect":false}}
     * ```
     * ## 数值输入框
     * ```json
     * {"title":"数值","field":"name","editor":{"controlType":EditorTypes.NUMBER,"precision":2,"step":1,"min":1,"max":10}}
     * ```
     * @param {?} columns
     * @param {?=} options
     * @return {?}
     */
    BatchEditDialogService.prototype.batchEdit = /**
     * \@description 控件参数格式
     * # 控件参数格式
     * ## textbox
     * ```json
     * {"title":"姓名","field":"name","editor":{"controlType":EditorTypes.TEXTBOX}}
     * ```
     * ## dropdown
     * ```json
     * {"title":"姓名","field":"name","editor":{"controlType":EditorTypes.COMBOLIST,"idField":"id","textField":"text","items":[],"multiSelect":false}}
     * ```
     * ## lookup
     * ```json
     * {"title":"姓名","field":"name","editor":{"controlType":EditorTypes.LOOKUP,"uri":"","idField":"id","textField":"text","valueField":"id","displayType":"LIST","mapFields":{},"singleSelect":true,"dialogTitle":"","hiddenInputName":"","items":[],"multiSelect":false}}
     * ```
     * ## 数值输入框
     * ```json
     * {"title":"数值","field":"name","editor":{"controlType":EditorTypes.NUMBER,"precision":2,"step":1,"min":1,"max":10}}
     * ```
     * @param {?} columns
     * @param {?=} options
     * @return {?}
     */
    function (columns, options) {
        var _this = this;
        /** @type {?} */
        var titleText = this.localeService.getValue('batchEditDialog.title');
        /** @type {?} */
        var okText = this.localeService.getValue('batchEditDialog.okText');
        /** @type {?} */
        var cancelText = this.localeService.getValue('batchEditDialog.cancelText');
        var _a = options || {}, _b = _a.width, width = _b === void 0 ? 792 : _b, _c = _a.height, height = _c === void 0 ? 580 : _c, _d = _a.title, title = _d === void 0 ? titleText : _d, _e = _a.rows, rows = _e === void 0 ? 1 : _e;
        /** @type {?} */
        var componentFactory = this.componentFactoryResolver.resolveComponentFactory(BatchEditDialogComponent);
        this.componentRef = componentFactory.create(this.injector);
        this.componentRef.instance.fields = this.getFields(columns);
        this.componentRef.instance.eidtors = this.getEditors(columns);
        this.componentRef.instance.title = title;
        this.componentRef.instance.rows = rows;
        this.dialogRef = this.modalService.show(this.componentRef, {
            title: title,
            width: width,
            height: height,
            buttons: [
                {
                    text: cancelText, cls: 'btn btn-secondary', handle: (/**
                     * @return {?}
                     */
                    function () {
                        _this.dialogRef.close();
                    })
                },
                {
                    text: okText, cls: 'btn btn-primary',
                    handle: (/**
                     * @return {?}
                     */
                    function () {
                        _this.userConfired();
                    })
                }
            ],
            showHeader: false,
            showButtons: false,
        });
        this.componentRef.instance.confirmed.subscribe((/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            _this.canConfirm(options).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                if (result === true) {
                    _this.userConfired(options);
                    _this.dialogRef.close();
                }
            }));
        }));
        this.componentRef.instance.canceled.subscribe((/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            _this.dialogRef.close();
        }));
        this.componentRef.instance.dialogRef = this.dialogRef;
        this.dialogRef.dialog.instance.draggbar.handle = this.componentRef.instance.header.nativeElement;
        return this.dialogRef;
    };
    /**
     * 用户确认修改
     */
    /**
     * 用户确认修改
     * @param {?=} options
     * @return {?}
     */
    BatchEditDialogService.prototype.userConfired = /**
     * 用户确认修改
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        /** @type {?} */
        var inputs = this.componentRef && this.componentRef.instance && this.componentRef.instance.inputs || [];
        // 针对checkbox做特殊处理
        if (inputs && inputs.length > 0) {
            inputs.forEach((/**
             * @param {?} input
             * @return {?}
             */
            function (input) {
                var controlType = input.controlType, dataType = input.dataType;
                if (controlType === 'checkbox' && dataType === 'boolean') {
                    input.value = input.value ? true : false;
                }
            }));
        }
        if (options && options.onConfirm && typeof options.onConfirm === 'function') {
            options.onConfirm(inputs);
        }
        this.onUserConfirmed.next(inputs);
    };
    /**
     * 从dom中获取grid列
     * @param columns dom
     */
    /**
     * 从dom中获取grid列
     * @private
     * @param {?} columns dom
     * @return {?}
     */
    BatchEditDialogService.prototype.getFields = /**
     * 从dom中获取grid列
     * @private
     * @param {?} columns dom
     * @return {?}
     */
    function (columns) {
        /** @type {?} */
        var fields = [];
        if (!columns || columns.length < 1) {
            return fields;
        }
        columns.forEach((/**
         * @param {?} collect
         * @return {?}
         */
        function (collect) {
            // collect为数组
            if (collect && collect.length > 0) {
                collect.forEach((/**
                 * @param {?} column
                 * @return {?}
                 */
                function (column) {
                    var _a = column || {}, _b = _a.field, field = _b === void 0 ? null : _b, _c = _a.title, title = _c === void 0 ? '' : _c, _d = _a.dataType, dataType = _d === void 0 ? null : _d, _e = _a.enableBatchEdit, enableBatchEdit = _e === void 0 ? false : _e;
                    if (field && enableBatchEdit) {
                        fields.push({ id: field, label: title, dataType: dataType });
                    }
                }));
            }
        }));
        return fields;
    };
    /**
     * 从columns中获取编辑器
     * @param columns dom
     */
    /**
     * 从columns中获取编辑器
     * @private
     * @param {?} columns dom
     * @return {?}
     */
    BatchEditDialogService.prototype.getEditors = /**
     * 从columns中获取编辑器
     * @private
     * @param {?} columns dom
     * @return {?}
     */
    function (columns) {
        /** @type {?} */
        var editors = [];
        if (!columns || columns.length < 1) {
            return editors;
        }
        columns.forEach((/**
         * @param {?} collect
         * @return {?}
         */
        function (collect) {
            // collect为数组
            if (collect && collect.length > 0) {
                collect.forEach((/**
                 * @param {?} column
                 * @return {?}
                 */
                function (column) {
                    var _a = column || {}, _b = _a.field, field = _b === void 0 ? null : _b, _c = _a.editor, editor = _c === void 0 ? {} : _c, _d = _a.enableBatchEdit, enableBatchEdit = _d === void 0 ? false : _d;
                    if (editor && field && enableBatchEdit) {
                        editor.controlType = editor.type;
                        editors.push({ field: field, editor: editor });
                    }
                }));
            }
        }));
        return editors;
    };
    /**
     * @private
     * @param {?=} options
     * @return {?}
     */
    BatchEditDialogService.prototype.canConfirm = /**
     * @private
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        /** @type {?} */
        var subject = new Subject();
        /** @type {?} */
        var showNotify = options && options.showNotify;
        var _a = (options || {}).rows, rows = _a === void 0 ? 1 : _a;
        /** @type {?} */
        var valueChanged = this.isValueChanged();
        if (!valueChanged) {
            return of(true);
        }
        if (showNotify === undefined) {
            showNotify = true;
        }
        if (!showNotify) {
            return of(true);
        }
        /** @type {?} */
        var dialogRef = null;
        try {
            /** @type {?} */
            var componentFactory = this.componentFactoryResolver.resolveComponentFactory(ResultConfirmComponent);
            /** @type {?} */
            var componentRef = componentFactory.create(this.injector);
            /** @type {?} */
            var shouldShowNotify = componentRef.instance.showNotify || false;
            componentRef.instance.rows = rows;
            if (shouldShowNotify) {
                dialogRef = this.modalService.show(componentRef, {
                    title: '',
                    width: 500,
                    height: 240,
                    showHeader: false,
                    showButtons: false,
                    buttons: []
                });
                // 确认
                componentRef.instance.confirmed.subscribe((/**
                 * @param {?} e
                 * @return {?}
                 */
                function (e) {
                    if (dialogRef) {
                        dialogRef.close();
                    }
                    subject.next(true);
                }));
                // 取消
                componentRef.instance.canceled.subscribe((/**
                 * @param {?} e
                 * @return {?}
                 */
                function (e) {
                    if (dialogRef) {
                        dialogRef.close();
                    }
                    subject.next(false);
                }));
            }
            else {
                return of(true);
            }
        }
        catch (e) {
            return of(true);
        }
        return subject;
    };
    /**
     * @return {?}
     */
    BatchEditDialogService.prototype.isValueChanged = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var inputs = this.componentRef && this.componentRef.instance && this.componentRef.instance.inputs || [];
        if (!inputs || inputs.length < 1) {
            return false;
        }
        /** @type {?} */
        var unChanged = inputs.every((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return !item.field; }));
        return !unChanged;
    };
    BatchEditDialogService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    BatchEditDialogService.ctorParameters = function () { return [
        { type: Injector },
        { type: BsModalService },
        { type: ComponentFactoryResolver },
        { type: LocaleService }
    ]; };
    return BatchEditDialogService;
}());
export { BatchEditDialogService };
if (false) {
    /** @type {?} */
    BatchEditDialogService.prototype.dialogRef;
    /**
     * @type {?}
     * @private
     */
    BatchEditDialogService.prototype.componentRef;
    /** @type {?} */
    BatchEditDialogService.prototype.onUserConfirmed;
    /** @type {?} */
    BatchEditDialogService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    BatchEditDialogService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    BatchEditDialogService.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    BatchEditDialogService.prototype.localeService;
}
