// @farris/ui-batch-edit-dialog/esm5/lib/lookup.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { LookupComponent } from '@farris/ui-lookup';
import { of } from 'rxjs';
var LookupDirective = /** @class */ (function () {
    function LookupDirective() {
        this.onSelectRows = new EventEmitter();
    }
    /**
     * @return {?}
     */
    LookupDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.lookup) {
            this.lookup.dictPicked = (/**
             * @param {?} rows
             * @return {?}
             */
            function (rows) {
                if (_this.onSelectRows) {
                    _this.onSelectRows.emit({ field: _this.field, rows: [].concat(rows) });
                }
                /** @type {?} */
                var result = {
                    closeDialog: true
                };
                return of(result);
            });
        }
    };
    LookupDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[lookup-extra]'
                },] }
    ];
    /** @nocollapse */
    LookupDirective.ctorParameters = function () { return []; };
    LookupDirective.propDecorators = {
        field: [{ type: Input, args: ['lookup-extra',] }],
        lookup: [{ type: Input, args: ['ins',] }],
        onSelectRows: [{ type: Output }]
    };
    return LookupDirective;
}());
export { LookupDirective };
if (false) {
    /** @type {?} */
    LookupDirective.prototype.field;
    /** @type {?} */
    LookupDirective.prototype.lookup;
    /** @type {?} */
    LookupDirective.prototype.onSelectRows;
}
