// @gspwidget/widget-core/esm5/lib/filter/filter.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, Pipe } from '@angular/core';
import { FilterSchema, FSEnum } from '@gspwidget/widget-devkit';
import { TranslateService } from '@ngx-translate/core';
var FilterComponent = /** @class */ (function () {
    function FilterComponent(translate) {
        this.translate = translate;
        this.singleValueChange = new EventEmitter();
        this.multiValueChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    FilterComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    FilterComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.fs && changes.fs.currentValue != changes.fs.previousValue) {
            this.value = FilterSchema.getDefaultValue(this.fs);
        }
    };
    /**
     * @return {?}
     */
    FilterComponent.prototype.getEnumItems = /**
     * @return {?}
     */
    function () {
        return FSEnum.GetItems(this.fs.enum);
    };
    /**
     * filter 不关心外界如何看待、处理多个值（多个值对应一个还是多个 dp），
     * 仅把多个值的数组发出去
     */
    /**
     * filter 不关心外界如何看待、处理多个值（多个值对应一个还是多个 dp），
     * 仅把多个值的数组发出去
     * @param {?} value
     * @return {?}
     */
    FilterComponent.prototype.handleChange = /**
     * filter 不关心外界如何看待、处理多个值（多个值对应一个还是多个 dp），
     * 仅把多个值的数组发出去
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.singleValueChange.emit(value);
    };
    /**
     * @param {?} values
     * @return {?}
     */
    FilterComponent.prototype.handleChangeMulti = /**
     * @param {?} values
     * @return {?}
     */
    function (values) {
        this.multiValueChange.emit(values);
    };
    FilterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'filter',
                    template: "<!-- {{fs.type|json}} -->\n<div class=\"d-flex align-items-center\" *ngIf=\"fs\">\n  <span *ngIf=\"fs.showName\"\n    class=\"text-truncate mr-1\">{{fs.i18nName[translate.currentLang]}}</span>\n  <ng-container *ngIf=\"fs.type==='date'\">\n    <ng-container *ngIf=\"fs.date.type==='single'\">\n      <nz-year-picker\n        *ngIf=\"fs.date.singleUnit==='year'\"\n        [(ngModel)]=\"value\"\n        nzFormat=\"YYYY\"\n        (ngModelChange)=\"handleChange($event)\"\n        ></nz-year-picker>\n      <nz-month-picker\n        *ngIf=\"fs.date.singleUnit==='month'\"\n        [(ngModel)]=\"value\"\n        nzFormat=\"YYYY-MM\"\n        (ngModelChange)=\"handleChange($event)\"\n        ></nz-month-picker>\n      <nz-date-picker\n        *ngIf=\"fs.date.singleUnit==='day'\"\n        [(ngModel)]=\"value\"\n        (ngModelChange)=\"handleChange($event)\"\n        ></nz-date-picker>\n    </ng-container>\n\n    <nz-range-picker\n      *ngIf=\"fs.date.type==='range'\"\n      [(ngModel)]=\"value\"\n      (ngModelChange)=\"handleChangeMulti($event)\"\n      ></nz-range-picker>\n\n    <ng-container *ngIf=\"fs.date.type=='lastTimeEnum'\">\n      <nz-select\n        class=\"\"\n        nzAllowClear\n        [(ngModel)]=\"value\"\n        (ngModelChange)=\"handleChange($event)\">\n        <nz-option\n          *ngFor=\"let lastTimeItem of fs.date.lastTimeItems\"\n          [nzValue]=\"lastTimeItem.num\"\n          [nzLabel]=\"\">last{{lastTimeItem.num}}{{lastTimeItem.unit}}</nz-option>\n      </nz-select>\n    </ng-container>\n  </ng-container>\n\n\n  <ng-container *ngIf=\"fs.type==='enum'\">\n    <!-- -{{getEnumItems()}}- -->\n    <!-- single -->\n    <nz-select\n      *ngIf=\"fs.enum.displayType==='select'\"\n      [nzMode]=\"fs.enum.multi ? 'multiple' : 'default'\"\n      class=\"col-sm-5\"\n      nzAllowClear\n      [(ngModel)]=\"value\"\n      (ngModelChange)=\"fs.enum.multi ? handleChangeMulti($event) : handleChange($event)\">\n      <nz-option\n        *ngFor=\"let item of getEnumItems()\"\n        [nzValue]=\"item.value\"\n        [nzLabel]=\"item.label || item.value\"></nz-option>\n    </nz-select>\n    <flat-select\n      *ngIf=\"!fs.enum.multi && fs.enum.displayType==='flat'\"\n      [options]=\"getEnumItems()\"\n      [(value)]=\"value\"\n      (valueChange)=\"handleChange($event)\"\n      ></flat-select>\n  </ng-container>\n\n\n  <ng-container *ngIf=\"fs.type==='number'\">\n    <input\n      type=\"number\"\n      class=\"form-control form-control-sm\"\n      [(ngModel)]=\"value\"\n      (ngModelChange)=\"handleChange($event)\"\n      required />\n  </ng-container>\n\n\n  <ng-container *ngIf=\"fs.type==='text'\">\n    <input\n      type=\"text\"\n      class=\"form-control form-control-sm\"\n      [(ngModel)]=\"value\"\n      (ngModelChange)=\"handleChange($event)\"\n      required />\n  </ng-container>\n</div>\n<ng-template #tagPlaceHolder let-selectedList> \u5DF2\u9009 {{selectedList.length}} \u9879</ng-template>\n",
                    styles: ["nz-select{min-width:60px}"]
                }] }
    ];
    /** @nocollapse */
    FilterComponent.ctorParameters = function () { return [
        { type: TranslateService }
    ]; };
    FilterComponent.propDecorators = {
        fs: [{ type: Input, args: ['filterSchema',] }],
        singleValueChange: [{ type: Output }],
        multiValueChange: [{ type: Output }]
    };
    return FilterComponent;
}());
export { FilterComponent };
if (false) {
    /**
     * filter 里绝不应该需要完整数据，否则参与查询的筛选还有啥意义？
     * @type {?}
     */
    FilterComponent.prototype.fs;
    /** @type {?} */
    FilterComponent.prototype.singleValueChange;
    /** @type {?} */
    FilterComponent.prototype.multiValueChange;
    /** @type {?} */
    FilterComponent.prototype.value;
    /** @type {?} */
    FilterComponent.prototype.translate;
}
var ColValuesPipe = /** @class */ (function () {
    function ColValuesPipe() {
    }
    /**
     * @param {?} data
     * @param {?} col
     * @return {?}
     */
    ColValuesPipe.prototype.transform = /**
     * @param {?} data
     * @param {?} col
     * @return {?}
     */
    function (data, col) {
        if (data && data.map)
            return Array.from(new Set(data.map((/**
             * @param {?} i
             * @return {?}
             */
            function (i) { return i[col]; }))));
    };
    ColValuesPipe.decorators = [
        { type: Pipe, args: [{ name: 'colValues' },] }
    ];
    return ColValuesPipe;
}());
export { ColValuesPipe };
