// @gsp-aif/runtime-api/esm5/lib/aif-documents-track/service/aif-documents-track.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, ComponentFactoryResolver } from "@angular/core";
import { HttpService } from "@ecp-caf/caf-common";
import { TrackType } from "../config/track-type.map";
import { AifDocumentsTrackManagerComponent } from "../aif-documents-track-manager/aif-documents-track-manager.component";
import { BsModalService } from '@farris/ui-modal';
import { AifDocumentsTrackManagerModuleConfig } from '../../entity/aif-documents-track-manager-module-config';
/**
 * 供内部和外部共同使用的单据追踪服务类，可输入参数自动查询并显示单据追踪模态框，也可以仅获取单据追踪查询结果
 */
var AifDocumentsTrackService = /** @class */ (function () {
    function AifDocumentsTrackService(http, modalService, injector, cfr, config) {
        this.http = http;
        this.modalService = modalService;
        this.injector = injector;
        this.cfr = cfr;
        this.serverIP = "";
        this.baseUrl = '/api/runtime/aif/v1.0/associationcontroller/';
        this.serverIP = config == null || config.serverIP == null ? "" : config.serverIP;
    }
    /**
     * 根据参数追踪并显示追踪模态框，注意：调用该方法时应同时引用TrackType枚举类
     * @param documentId 单据ID
     * @param beType 单据类型编号
     * @param trackType 追踪类型，仅可用“UpSearch”（上查）或“DownSearch”（下查)
     */
    /**
     * 根据参数追踪并显示追踪模态框，注意：调用该方法时应同时引用TrackType枚举类
     * @param {?} documentId 单据ID
     * @param {?} beType 单据类型编号
     * @param {?} trackType 追踪类型，仅可用“UpSearch”（上查）或“DownSearch”（下查)
     * @return {?}
     */
    AifDocumentsTrackService.prototype.showResultDialog = /**
     * 根据参数追踪并显示追踪模态框，注意：调用该方法时应同时引用TrackType枚举类
     * @param {?} documentId 单据ID
     * @param {?} beType 单据类型编号
     * @param {?} trackType 追踪类型，仅可用“UpSearch”（上查）或“DownSearch”（下查)
     * @return {?}
     */
    function (documentId, beType, trackType) {
        /** @type {?} */
        var cmpF = this.cfr.resolveComponentFactory(AifDocumentsTrackManagerComponent);
        /** @type {?} */
        var cmpR = cmpF.create(this.injector);
        if (trackType != TrackType.DownSearch && trackType != TrackType.UpSearch) {
            trackType = TrackType.DownSearch;
        }
        cmpR.instance.searchFromOut(documentId, beType, trackType);
        return this.modalService.show(cmpR, {
            title: '追踪',
            width: 1100,
            height: 600,
            buttons: null,
            showButtons: false
        });
    };
    /**
     * 根据参数追踪并将查询结果返回，注意：调用该方法时应同时引用TrackType枚举类
     * @param documentId 单据ID
     * @param beType 单据类型编号
     * @param trackType 追踪类型，仅可用“UpSearch”（上查）或“DownSearch”（下查),
     */
    /**
     * 根据参数追踪并将查询结果返回，注意：调用该方法时应同时引用TrackType枚举类
     * @param {?} documentId 单据ID
     * @param {?} beType 单据类型编号
     * @param {?} trackType 追踪类型，仅可用“UpSearch”（上查）或“DownSearch”（下查),
     * @return {?}
     */
    AifDocumentsTrackService.prototype.getTrackedDocumentData = /**
     * 根据参数追踪并将查询结果返回，注意：调用该方法时应同时引用TrackType枚举类
     * @param {?} documentId 单据ID
     * @param {?} beType 单据类型编号
     * @param {?} trackType 追踪类型，仅可用“UpSearch”（上查）或“DownSearch”（下查),
     * @return {?}
     */
    function (documentId, beType, trackType) {
        if (trackType == TrackType.DownSearch) {
            return this.getTargetData(beType, documentId);
        }
        else {
            return this.getSourceData(beType, documentId);
        }
    };
    /**
     * @private
     * @param {?} srcBeType
     * @param {?} srcDataId
     * @return {?}
     */
    AifDocumentsTrackService.prototype.getTargetData = /**
     * @private
     * @param {?} srcBeType
     * @param {?} srcDataId
     * @return {?}
     */
    function (srcBeType, srcDataId) {
        /** @type {?} */
        var url = this.serverIP + this.baseUrl + ("GetTargetData/" + srcBeType + "&" + srcDataId);
        return this.http.get(url);
    };
    /**
     * @private
     * @param {?} tarBeType
     * @param {?} tarDataId
     * @return {?}
     */
    AifDocumentsTrackService.prototype.getSourceData = /**
     * @private
     * @param {?} tarBeType
     * @param {?} tarDataId
     * @return {?}
     */
    function (tarBeType, tarDataId) {
        /** @type {?} */
        var url = this.serverIP + this.baseUrl + ("GetSourceData/" + tarBeType + "&" + tarDataId);
        return this.http.get(url);
    };
    AifDocumentsTrackService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AifDocumentsTrackService.ctorParameters = function () { return [
        { type: HttpService },
        { type: BsModalService },
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: AifDocumentsTrackManagerModuleConfig }
    ]; };
    return AifDocumentsTrackService;
}());
export { AifDocumentsTrackService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AifDocumentsTrackService.prototype.serverIP;
    /**
     * @type {?}
     * @private
     */
    AifDocumentsTrackService.prototype.baseUrl;
    /**
     * @type {?}
     * @private
     */
    AifDocumentsTrackService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    AifDocumentsTrackService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    AifDocumentsTrackService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    AifDocumentsTrackService.prototype.cfr;
}
