// @gsp-lcm/bo-creator/esm5/lib/views/bo-creator/bo-creator.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/views/bo-creator/bo-creator.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input, ViewChild } from '@angular/core';
import { AppCreatorModalComponent } from './modal/app-modal/app-creator-modal.component';
import { SuCreatorModalComponent } from './modal/su-modal/su-creator-modal.component';
import { BoCreatorModalComponent } from './modal/bo-modal/bo-creator-modal.component';
var BOCreatorComponent = /** @class */ (function () {
    function BOCreatorComponent() {
        this.createdAfterOk = new EventEmitter();
        this.modalClosed = new EventEmitter();
    }
    /**
     * @return {?}
     */
    BOCreatorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} $event
     * @return {?}
     */
    BOCreatorComponent.prototype.appConfirmEmitterHandler = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var bo = $event;
        this.createdAfterOk.emit(bo);
        this.modalClosed.emit();
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    BOCreatorComponent.prototype.suConfirmEmitterHandler = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var bo = $event;
        this.createdAfterOk.emit(bo);
        this.modalClosed.emit();
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    BOCreatorComponent.prototype.boConfirmEmitterHandler = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var bo = $event;
        this.createdAfterOk.emit(bo);
        this.modalClosed.emit();
    };
    /**
     * @return {?}
     */
    BOCreatorComponent.prototype.btnOkClick = /**
     * @return {?}
     */
    function () {
        switch (this.type) {
            case 'app':
                this.appCmp.create();
                return;
            case 'su':
                this.suCmp.create();
                return;
            case 'bo':
                this.boCmp.create();
                return;
            default:
                return;
        }
    };
    /**
     * @return {?}
     */
    BOCreatorComponent.prototype.hideModal = /**
     * @return {?}
     */
    function () {
        this.modalClosed.emit();
    };
    BOCreatorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-bo-creator',
                    template: "<div class=\"bo-creator farris-main-area flex-column farris-overflow-hidden\">\r\n\r\n  <div *ngIf=\"type==='app'\" class=\"modal-body boCreatorPanel\">\r\n    <app-app-creator-modal #appModal (appConfirmEmitter)=\"appConfirmEmitterHandler($event)\"></app-app-creator-modal>\r\n  </div>\r\n\r\n  <div *ngIf=\"type==='su'\" class=\"modal-body boCreatorPanel\">\r\n    <app-su-creator-modal #suModal (suConfirmEmitter)=\"suConfirmEmitterHandler($event)\" [(parentId)]=\"parentId\">\r\n    </app-su-creator-modal>\r\n  </div>\r\n\r\n  <div *ngIf=\"type ==='bo'\" class=\"modal-body boCreatorPanel\">\r\n    <app-bo-creator-modal #boModal (boConfirmEmitter)=\"boConfirmEmitterHandler($event)\" [(_parentId)]=\"parentId\">\r\n    </app-bo-creator-modal>\r\n  </div>\r\n\r\n  <div class=\"modal-footer btns\" #btns>\r\n    <button type=\"button\" class=\"btn btn-default btn-primary\" (click)=\"btnOkClick()\">\u786E\u5B9A</button>\r\n    <button type=\"button\" class=\"btn btn-default\" (click)=\"hideModal()\">\u53D6\u6D88</button>\r\n  </div>\r\n</div>",
                    styles: [".sucreator-wrapper{background:rgba(255,255,255,.8);width:95%;margin:40px auto 0;max-width:480px;padding:24px 18px;border-radius:8px;position:relative;z-index:100}.suInfo{margin:0 0 16px 20px}.sucreator-wrapper .col-form-label,.sucreator-wrapper .form-control{padding-top:3px;padding-bottom:3px}.sucreator-wrapper .col-form-label{width:70px;margin-right:12px;text-align:right;flex-shrink:0}.bocreator{background:#fff;box-shadow:0 2px 16px 0 rgba(0,0,0,.1);border-radius:4px}.boCreatorPanel{background-color:#fff}"]
                }] }
    ];
    /** @nocollapse */
    BOCreatorComponent.ctorParameters = function () { return []; };
    BOCreatorComponent.propDecorators = {
        createdAfterOk: [{ type: Output }],
        modalClosed: [{ type: Output }],
        parentId: [{ type: Input }],
        type: [{ type: Input }],
        appCmp: [{ type: ViewChild, args: ["appModal",] }],
        suCmp: [{ type: ViewChild, args: ["suModal",] }],
        boCmp: [{ type: ViewChild, args: ["boModal",] }]
    };
    return BOCreatorComponent;
}());
export { BOCreatorComponent };
if (false) {
    /** @type {?} */
    BOCreatorComponent.prototype.createdAfterOk;
    /** @type {?} */
    BOCreatorComponent.prototype.modalClosed;
    /** @type {?} */
    BOCreatorComponent.prototype.parentId;
    /** @type {?} */
    BOCreatorComponent.prototype.type;
    /** @type {?} */
    BOCreatorComponent.prototype.appCmp;
    /** @type {?} */
    BOCreatorComponent.prototype.suCmp;
    /** @type {?} */
    BOCreatorComponent.prototype.boCmp;
}
