// @farris/ui-button/esm5/lib/button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output, ViewChild, ElementRef, Optional, Injector } from '@angular/core';
var FarrisButtonComponent = /** @class */ (function () {
    function FarrisButtonComponent(injector, _elRef) {
        var _this = this;
        this.injector = injector;
        this._elRef = _elRef;
        this.size = 'small';
        this.type = 'primary';
        this.click = new EventEmitter();
        this.__zone_click = null;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.el = _elRef ? _elRef.nativeElement : _this.btnEl ? _this.btnEl.nativeElement.parentElement : null;
            _this.initButtonEvent();
        }));
    }
    /**
     * @private
     * @return {?}
     */
    FarrisButtonComponent.prototype.initButtonEvent = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.el && this.el.closest('grid-body-cell')) {
            this.el.addEventListener('mousedown', this.onBtnMouseDown.bind(this), true);
        }
    };
    /**
     * @private
     * @param {?} e
     * @return {?}
     */
    FarrisButtonComponent.prototype.onBtnMouseDown = /**
     * @private
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.__zone_click = this.el['__zone_symbol__clickfalse'];
        if (this.__zone_click && this.__zone_click.length) {
            e.currentTarget.style.pointerEvents = 'none';
            this.el['__zone_symbol__clickfalse'] = null;
        }
    };
    /**
     * @param {?} e
     * @return {?}
     */
    FarrisButtonComponent.prototype.clickEvent = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        e.stopPropagation();
        if (this.disabled) {
            return;
        }
        this.click.emit(e);
    };
    FarrisButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-button',
                    template: "\n    <button [class]=\"(size === 'large'?'btn-lg ':'btn-sm ')+\n    'btn '+ (type?'btn-'+type+' ':'')\"  #btn \n    [class.disabled]=\"disabled\" (click)=\"clickEvent($event)\">\n      <ng-content></ng-content>\n    </button>\n  "
                }] }
    ];
    /** @nocollapse */
    FarrisButtonComponent.ctorParameters = function () { return [
        { type: Injector, decorators: [{ type: Optional }] },
        { type: ElementRef, decorators: [{ type: Optional }] }
    ]; };
    FarrisButtonComponent.propDecorators = {
        text: [{ type: Input }],
        size: [{ type: Input }],
        type: [{ type: Input }],
        disabled: [{ type: Input }],
        click: [{ type: Output }],
        btnEl: [{ type: ViewChild, args: ['btn',] }]
    };
    return FarrisButtonComponent;
}());
export { FarrisButtonComponent };
if (false) {
    /** @type {?} */
    FarrisButtonComponent.prototype.text;
    /** @type {?} */
    FarrisButtonComponent.prototype.size;
    /** @type {?} */
    FarrisButtonComponent.prototype.type;
    /** @type {?} */
    FarrisButtonComponent.prototype.disabled;
    /** @type {?} */
    FarrisButtonComponent.prototype.click;
    /** @type {?} */
    FarrisButtonComponent.prototype.btnEl;
    /** @type {?} */
    FarrisButtonComponent.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FarrisButtonComponent.prototype.__zone_click;
    /**
     * @type {?}
     * @private
     */
    FarrisButtonComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    FarrisButtonComponent.prototype._elRef;
}
