// @gsp-cmp/ubcc-param-data-ui/esm5/lib/ubcc-param-data-ui/ubcc-param-data-ui.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ubcc-param-data-ui/ubcc-param-data-ui.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
import { UbccParamTemplate } from '@gsp-cmp/ubcc-param-vo';
import { DataTypeConst } from '@ecp-caf/common-structure';
var UbccParamDataUiComponent = /** @class */ (function () {
    function UbccParamDataUiComponent() {
        this.paramDataItems = [];
        this.canEdit = true;
    }
    /**
     * @return {?}
     */
    UbccParamDataUiComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.template = this.template || new UbccParamTemplate();
        this.template.params = this.template.params || [];
        this.template.groups = this.template.groups || [];
    };
    /**
     * @param {?} param
     * @return {?}
     */
    UbccParamDataUiComponent.prototype.getControlType = /**
     * @param {?} param
     * @return {?}
     */
    function (param) {
        if (!param || !param.parameterType || !param.parameterTypeDesc) {
            return;
        }
        switch (param.parameterTypeDesc) {
            case 'string':
                return 'input';
            case 'bool':
                return 'checkbox';
            case 'double':
            case 'int':
                return 'number';
            case 'datetime':
                return 'date';
            case 'enum':
                return 'dropdownlist';
            case 'help':
                return 'lookup';
            default:
                return 'input';
        }
    };
    /**
     * @param {?} group
     * @return {?}
     */
    UbccParamDataUiComponent.prototype.getParamItems = /**
     * @param {?} group
     * @return {?}
     */
    function (group) {
        return this.paramDataItems.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.groupId === group.groupId; }));
    };
    /**
     * @param {?} group
     * @param {?} index
     * @return {?}
     */
    UbccParamDataUiComponent.prototype.getIndex = /**
     * @param {?} group
     * @param {?} index
     * @return {?}
     */
    function (group, index) {
        /** @type {?} */
        var currentGroupParams = this.getParamItems(group);
        /** @type {?} */
        var currentParam = this.paramDataItems.find((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.id === currentGroupParams[index].id; }));
        /** @type {?} */
        var realIndex = this.paramDataItems.findIndex((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.id === currentParam.id; }));
        return realIndex;
    };
    /**
     * @param {?} param
     * @return {?}
     */
    UbccParamDataUiComponent.prototype.show = /**
     * @param {?} param
     * @return {?}
     */
    function (param) {
        return param.visible || false;
    };
    /**
     * @param {?} param
     * @return {?}
     */
    UbccParamDataUiComponent.prototype.isCollection = /**
     * @param {?} param
     * @return {?}
     */
    function (param) {
        if (!param || !param.parameterType) {
            return false;
        }
        if (param.parameterType.getKind() === DataTypeConst.DEFAULT_COLLECTION) {
            return true;
        }
        else {
            return false;
        }
    };
    UbccParamDataUiComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-ubcc-param-data-ui',
                    template: "<farris-section [mainTitle]=\"'\u53C2\u6570\u4FE1\u606F'\" [enableMaximize]=\"false\">\r\n\r\n    <div *ngIf=\"!template.groupEnable\">\r\n        <div *ngFor=\"let parameter of paramDataItems;let i=index;\" style=\"margin-top: 10px;\">\r\n\r\n            <div *ngIf=\"show(paramDataItems[i])\">\r\n                <div class=\"col-12\" [ngSwitch]=\"getControlType(paramDataItems[i])\">\r\n\r\n                    <lib-dropdownlist-control *ngSwitchCase=\"'dropdownlist'\" [(canEdit)]=\"canEdit\" [(param)]=\"paramDataItems[i]\"\r\n                        [(template)]=\"template\"></lib-dropdownlist-control>\r\n\r\n                    <lib-lookup-control *ngSwitchCase=\"'lookup'\" [(canEdit)]=\"canEdit\" [(param)]=\"paramDataItems[i]\"></lib-lookup-control>\r\n\r\n                    <lib-number-control *ngSwitchCase=\"'number'\" [(canEdit)]=\"canEdit\" [(param)]=\"paramDataItems[i]\"></lib-number-control>\r\n\r\n                    <lib-date-control *ngSwitchCase=\"'date'\" [(canEdit)]=\"canEdit\" [(param)]=\"paramDataItems[i]\"></lib-date-control>\r\n\r\n                    <lib-checkbox-control *ngSwitchCase=\"'checkbox'\" [(canEdit)]=\"canEdit\" [(param)]=\"paramDataItems[i]\"></lib-checkbox-control>\r\n\r\n                    <lib-input-control *ngSwitchDefault [(canEdit)]=\"canEdit\" [(param)]=\"paramDataItems[i]\"></lib-input-control>\r\n\r\n                </div>\r\n            </div>\r\n        </div>\r\n    </div>\r\n\r\n    <div *ngIf=\"template.groupEnable\" style=\"margin-top: 10px;\">\r\n\r\n        <farris-panel [defaultExpand]=\"true\">\r\n            <farris-panel-item class=\"border-0\" *ngFor=\"let group of template.groups;let groupIndex=index;\" [title]=\"group.groupName\" style=\"padding-bottom: 2px;\">\r\n                <ng-template #contentTempl>\r\n                    <div *ngFor=\"let parameter of getParamItems(template.groups[groupIndex]);let i=index;\" style=\"margin-top: 10px;\">\r\n\r\n                        <div *ngIf=\"show(paramDataItems[getIndex(template.groups[groupIndex],i)])\">\r\n                            <div class=\"col-12\" [ngSwitch]=\"getControlType(paramDataItems[getIndex(template.groups[groupIndex],i)])\">\r\n\r\n                                <lib-dropdownlist-control *ngSwitchCase=\"'dropdownlist'\" [(canEdit)]=\"canEdit\"\r\n                                    [(param)]=\"paramDataItems[getIndex(template.groups[groupIndex],i)]\" [(template)]=\"template\"></lib-dropdownlist-control>\r\n\r\n                                <lib-lookup-control *ngSwitchCase=\"'lookup'\" [(canEdit)]=\"canEdit\" [(param)]=\"paramDataItems[getIndex(template.groups[groupIndex],i)]\"></lib-lookup-control>\r\n\r\n                                <lib-number-control *ngSwitchCase=\"'number'\" [(canEdit)]=\"canEdit\" [(param)]=\"paramDataItems[getIndex(template.groups[groupIndex],i)]\"></lib-number-control>\r\n\r\n                                <lib-date-control *ngSwitchCase=\"'date'\" [(canEdit)]=\"canEdit\" [(param)]=\"paramDataItems[getIndex(template.groups[groupIndex],i)]\"></lib-date-control>\r\n\r\n                                <lib-checkbox-control *ngSwitchCase=\"'checkbox'\" [(canEdit)]=\"canEdit\" [(param)]=\"paramDataItems[getIndex(template.groups[groupIndex],i)]\"></lib-checkbox-control>\r\n\r\n                                <lib-input-control *ngSwitchDefault [(canEdit)]=\"canEdit\" [(param)]=\"paramDataItems[getIndex(template.groups[groupIndex],i)]\"></lib-input-control>\r\n\r\n                            </div>\r\n                        </div>\r\n                    </div>\r\n                </ng-template>\r\n            </farris-panel-item>\r\n        </farris-panel>\r\n    </div>\r\n\r\n</farris-section>",
                    encapsulation: ViewEncapsulation.None,
                    styles: [":host{display:flex;flex-direction:column;flex:1}.farris-panel .card.f-state-selected .card-header,lib-ubcc-param-data-ui .accordion .card.f-state-selected .card-header{color:#262626!important;background-color:#f4f4f4!important;border-color:#e6e6e6}.farris-panel .card .card-header,lib-ubcc-param-data-ui .accordion .card .card-header{position:relative;border-width:0;background-color:#f4f4f4!important;color:rgba(0,0,0,.85);padding:.5265rem .875rem}.farris-panel .card .card-header .panel-item-title,lib-ubcc-param-data-ui .accordion .card .card-header .panel-item-title{font-size:15px!important;line-height:1.375rem}lib-ubcc-param-data-ui .farris-panel{border:0 solid rgba(0,0,0,.125)!important}.farris-panel .card.f-state-selected .card-header .f-icon,lib-ubcc-param-data-ui .accordion .card.f-state-selected .card-header .f-icon{color:#171616!important}.f-section-header,lib-ubcc-param-data-ui .f-section-extend{margin:0 0 .25rem}"]
                }] }
    ];
    /** @nocollapse */
    UbccParamDataUiComponent.ctorParameters = function () { return []; };
    UbccParamDataUiComponent.propDecorators = {
        template: [{ type: Input }],
        paramDataItems: [{ type: Input }],
        canEdit: [{ type: Input }]
    };
    return UbccParamDataUiComponent;
}());
export { UbccParamDataUiComponent };
if (false) {
    /** @type {?} */
    UbccParamDataUiComponent.prototype.template;
    /** @type {?} */
    UbccParamDataUiComponent.prototype.paramDataItems;
    /** @type {?} */
    UbccParamDataUiComponent.prototype.canEdit;
}
