// @farris/ui-lookup/esm2015/lib/lookup-tip.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Renderer2, Injector, Input } from '@angular/core';
import { InputGroupComponent } from '@farris/ui-input-group';
import { PersonalConfigService } from './lookup-presonalConfig.service';
import { delay } from 'rxjs/operators';
import { LocaleService } from '@farris/ui-locale';
export class LookupTipDirective {
    /**
     * @param {?} inputRef
     * @param {?} renderer
     * @param {?} injector
     */
    constructor(inputRef, renderer, injector) {
        this.inputRef = inputRef;
        this.renderer = renderer;
        this.injector = injector;
        this.enableTip = false;
        this.isInTipPanel = false;
        this.tipText = '您要找的是不是这些？';
        this.personalConfigService = this.injector.get(PersonalConfigService, null);
        this.localService = this.injector.get(LocaleService);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.tipText = this.localService.getValue('lookup.tipText');
        this.enableTip = false; // 因数据权限问题，此功能暂屏蔽 tfs: 403924
        if (this.enableTip) {
            this.inputRef.iconMouseEnter.subscribe((/**
             * @param {?} e
             * @return {?}
             */
            e => {
                this.creatPanel();
            }));
            this.inputRef.iconMouseLeave.pipe(delay(200)).subscribe((/**
             * @param {?} e
             * @return {?}
             */
            e => {
                if (!this.isInTipPanel) {
                    this.removePanel();
                }
            }));
            this.inputRef.clickHandle.subscribe((/**
             * @param {?} e
             * @return {?}
             */
            e => {
                this.removePanel();
            }));
        }
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() { }
    // 获取快捷输入数据
    /**
     * @private
     * @return {?}
     */
    getQuickSelectedByLocaleId() {
        /** @type {?} */
        const localeid = this.localService.localeId;
        /** @type {?} */
        const personalConf = this.personalConfigService.getPersonalData();
        if (personalConf && personalConf.quickSelected) {
            /** @type {?} */
            const items = personalConf.quickSelected[localeid];
            if (items && items.length) {
                return items;
            }
        }
        return null;
    }
    /**
     * @return {?}
     */
    creatPanel() {
        this.selected = this.getQuickSelectedByLocaleId();
        if (this.tipPanel || !this.selected || (this.selected && this.selected.length === 0)) {
            return;
        }
        /** @type {?} */
        const pos = this.setPosition();
        /** @type {?} */
        const ul = this.renderer.createElement('ul');
        /** @type {?} */
        const path = this.personalConfigService.textField.split('.');
        this.selected.forEach((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        (item, index) => {
            /** @type {?} */
            const li = this.renderer.createElement('li');
            li.innerHTML = this.find(item, this.personalConfigService.textField);
            this.renderer.setProperty(li, 'id', index);
            this.renderer.appendChild(ul, li);
        }));
        /** @type {?} */
        const header = this.renderer.createElement('div');
        this.renderer.addClass(header, 'lookup-tip-header');
        header.innerHTML = this.tipText;
        this.tipPanel = this.renderer.createElement('div');
        this.renderer.appendChild(this.tipPanel, header);
        this.renderer.appendChild(this.tipPanel, ul);
        this.renderer.addClass(this.tipPanel, 'lookup-tip');
        this.renderer.setStyle(this.tipPanel, 'top', pos.top);
        this.renderer.setStyle(this.tipPanel, 'left', pos.left);
        this.renderer.setStyle(this.tipPanel, 'width', pos.width);
        this.renderer.appendChild(document.body, this.tipPanel);
        this.leaveEvent = this.renderer.listen(this.tipPanel, 'mouseleave', (/**
         * @return {?}
         */
        () => {
            this.removePanel();
        }));
        this.enterEvent = this.renderer.listen(this.tipPanel, 'mouseenter', (/**
         * @return {?}
         */
        () => {
            this.isInTipPanel = true;
        }));
        this.clickEvent = this.renderer.listen(this.tipPanel, 'click', (/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            if (e.target.nodeName === 'LI') {
                this.selectItem(e.target.id);
            }
        }));
    }
    /**
     * @return {?}
     */
    removePanel() {
        if (document.body.contains(this.tipPanel)) {
            this.renderer.removeChild(document.body, this.tipPanel);
            this.tipPanel = null;
            this.isInTipPanel = false;
            this.leaveEvent();
            this.enterEvent();
            this.clickEvent();
        }
    }
    /**
     * @return {?}
     */
    setPosition() {
        const { left, top, width, height } = this.inputRef.inputGroup.nativeElement.getBoundingClientRect();
        return {
            left: left + 'px',
            top: top + height + 'px',
            width: width + 'px'
        };
    }
    /**
     * @param {?} val
     * @return {?}
     */
    selectItem(val) {
        /** @type {?} */
        const item = this.selected.find((/**
         * @param {?} el
         * @param {?} index
         * @return {?}
         */
        (el, index) => Number(val) === index));
        this.removePanel();
        if (this.personalConfigService.singleSelect) {
            this.personalConfigService.selectItemObser$.next(item);
        }
        else {
            this.personalConfigService.selectItemObser$.next([item]);
        }
    }
    /**
     * @param {?} obj
     * @param {?} keys
     * @return {?}
     */
    find(obj, keys) {
        /** @type {?} */
        const keyArr = keys.split('.');
        /** @type {?} */
        const key = keyArr[0];
        /** @type {?} */
        const target = obj[key];
        if (target instanceof Object) {
            keyArr.shift();
            return this.find(target, keyArr.join('.'));
        }
        else {
            return target;
        }
    }
}
LookupTipDirective.decorators = [
    { type: Directive, args: [{
                selector: '[lookup-tip]'
            },] }
];
/** @nocollapse */
LookupTipDirective.ctorParameters = () => [
    { type: InputGroupComponent },
    { type: Renderer2 },
    { type: Injector }
];
LookupTipDirective.propDecorators = {
    enableTip: [{ type: Input, args: ['lookup-tip',] }]
};
if (false) {
    /** @type {?} */
    LookupTipDirective.prototype.enableTip;
    /**
     * @type {?}
     * @private
     */
    LookupTipDirective.prototype.tipPanel;
    /**
     * @type {?}
     * @private
     */
    LookupTipDirective.prototype.leaveEvent;
    /**
     * @type {?}
     * @private
     */
    LookupTipDirective.prototype.enterEvent;
    /**
     * @type {?}
     * @private
     */
    LookupTipDirective.prototype.clickEvent;
    /**
     * @type {?}
     * @private
     */
    LookupTipDirective.prototype.isInTipPanel;
    /**
     * @type {?}
     * @private
     */
    LookupTipDirective.prototype.selected;
    /**
     * @type {?}
     * @private
     */
    LookupTipDirective.prototype.personalConfigService;
    /** @type {?} */
    LookupTipDirective.prototype.localService;
    /**
     * @type {?}
     * @private
     */
    LookupTipDirective.prototype.tipText;
    /**
     * @type {?}
     * @private
     */
    LookupTipDirective.prototype.inputRef;
    /**
     * @type {?}
     * @private
     */
    LookupTipDirective.prototype.renderer;
    /** @type {?} */
    LookupTipDirective.prototype.injector;
}
