// @gsp-lcm/metadata-selector/esm2015/metadata/metadata-select/components/metadatalist/metadatalist.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: metadata/metadata-select/components/metadatalist/metadatalist.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output } from '@angular/core';
import { MetadataService } from '../../metadata.service';
import { GSPMetadataService } from '../../../metadata.service';
import { distinctUntilChanged, debounceTime } from 'rxjs/operators';
import { Subject } from 'rxjs';
export class TemplatemetadataComponent {
    /**
     * @param {?} metadataService
     * @param {?} gspMetaService
     */
    constructor(metadataService, gspMetaService) {
        this.metadataService = metadataService;
        this.gspMetaService = gspMetaService;
        this.subscriptions = [];
        this.dotnet = 'dotnet';
        this.java = 'java';
        this.technologyStack = new EventEmitter();
        this.loading = false;
        this.selectedMetadataChanged = new EventEmitter();
        this.allSource = { text: '全选', value: 'All' };
        this.selectedSource = this.allSource;
        this.sourceList = [this.allSource];
        this.popupSettings = { width: 'auto' };
        this.allowCustom = true;
        this.allType = { text: '全选', value: 'All' };
        this.selectedType = this.allType;
        this.typeList = [
            this.allType
        ];
        this.postfixSelected = '';
        this.initPageIndex = 0;
        this.initPageSize = 10;
        this.page = {
            pageIndex: this.initPageIndex,
            pageSize: this.initPageSize
        };
        this.typeListReadonly = false;
        this.hasNext = true;
        this.searchText = '';
        this.searchTextForDup = '';
        this.searchTextSub = new Subject();
        this.postfix = '';
        this.postfixFlag = true;
        this.scope = 0;
        this.conflictAvoidFlag = false;
        this.reload = true;
        this.mySelection = [];
        // tslint:disable-next-line:member-ordering
        this.state = {
            skip: 0,
            take: 9999,
            // Initial filter descriptor
            filter: {
                logic: 'and',
                filters: []
            }
        };
        // tslint:disable-next-line:member-ordering
        this.selectableSetting = { mode: 'single', checkboxOnly: false };
        this.rowSelected = (/**
         * @param {?} e
         * @return {?}
         */
        (e) => this.mySelection.indexOf(e.index) >= 0);
        this.classifier = { id: '' };
        this.searchTextSub.pipe(debounceTime(2048), distinctUntilChanged()).subscribe((/**
         * @param {?} text
         * @return {?}
         */
        text => {
            this.searchText = text.trim();
            // 搜索
            console.log('search for ' + this.searchText);
            this.getData();
        }));
        // this.unsubscribeAll();
        /** @type {?} */
        const subscription = this.metadataService.selectedChange.subscribe((/**
         * @param {?} classifier
         * @return {?}
         */
        classifier => {
            if (this.classifier && classifier && this.classifier.id === classifier.id) {
                return;
            }
            // 元数据类型
            if (this.postfixFlag) {
                this.gspMetaService.GetMetadataTypeList().subscribe((/**
                 * @param {?} typeList
                 * @return {?}
                 */
                typeList => {
                    /** @type {?} */
                    let count = 0;
                    typeList.forEach((/**
                     * @param {?} type
                     * @return {?}
                     */
                    type => {
                        this.postfixFlag = false;
                        /** @type {?} */
                        const typeNameArray = this.typeName.split(',');
                        typeNameArray.forEach((/**
                         * @param {?} typeName
                         * @return {?}
                         */
                        typeName => {
                            if (typeName.toLowerCase() === type.typeCode.toLowerCase()) {
                                this.postfix = this.postfix + type.postfix + ',';
                                this.typeList.push({ text: type.typeName, value: type.postfix });
                                count++;
                            }
                        }));
                    }));
                    if (count === 1) {
                        this.typeList.shift();
                        this.selectedType = this.typeList[0];
                        this.typeListReadonly = true;
                    }
                    this.postfixSelected = this.postfix;
                    this.scopeChanged(classifier);
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                error => {
                    /** @type {?} */
                    const msg = error.error.Message !== undefined ? error.error.Message : error.error.error.message;
                    gsp.ide.messager.error('Fetch Metadata Type error:' + msg);
                }));
            }
            else {
                this.scopeChanged(classifier);
            }
            // 技术栈赋值
            this.metadataService.getCodeLanguage().subscribe((/**
             * @param {?} item
             * @return {?}
             */
            item => {
                if (item) {
                    // 如果目前工程非单栈，则获取环境默认技术栈
                    if (item.indexOf(';') > -1) {
                        this.gspMetaService.getSettingInfo().subscribe((/**
                         * @param {?} setting
                         * @return {?}
                         */
                        setting => {
                            if (setting && setting.technologyStack) {
                                this.codelanguage = setting.technologyStack;
                                this.technologyStack.emit(this.codelanguage);
                                if (classifier.id === 'ref' || classifier.id === 'remote') {
                                    if (this.codelanguage === 'dotnet' || this.isMavenEnable === false) {
                                        this.initNugetSourceList();
                                    }
                                    else {
                                        this.initMavenSourceList();
                                    }
                                }
                            }
                            else if (this.isMavenEnable === true) {
                                // 环境默认配置为空，则按照java技术栈处理
                                this.codelanguage = 'java';
                                this.technologyStack.emit(this.codelanguage);
                                if (classifier.id === 'ref' || classifier.id === 'remote') {
                                    this.initMavenSourceList();
                                }
                            }
                            else {
                                this.codelanguage = 'dotnet';
                                this.technologyStack.emit(this.codelanguage);
                                if (classifier.id === 'ref' || classifier.id === 'remote') {
                                    this.initNugetSourceList();
                                }
                            }
                        }));
                    }
                    else {
                        this.codelanguage = item;
                        this.technologyStack.emit(this.codelanguage);
                        if (classifier.id === 'ref' || classifier.id === 'remote') {
                            if (this.codelanguage === 'dotnet' || this.isMavenEnable === false) {
                                this.initNugetSourceList();
                            }
                            else {
                                this.initMavenSourceList();
                            }
                        }
                    }
                }
                else {
                    // 工程的技术栈为空，则也是从环境默认技术栈读取
                    this.gspMetaService.getSettingInfo().subscribe((/**
                     * @param {?} setting
                     * @return {?}
                     */
                    setting => {
                        if (setting && setting.technologyStack) {
                            this.codelanguage = setting.technologyStack;
                            this.technologyStack.emit(this.codelanguage);
                            if (classifier.id === 'ref' || classifier.id === 'remote') {
                                if (this.codelanguage === 'dotnet' || this.isMavenEnable === false) {
                                    this.initNugetSourceList();
                                }
                                else {
                                    this.initMavenSourceList();
                                }
                            }
                        }
                        else if (this.isMavenEnable === true) {
                            // 环境默认配置为空，则按照java技术栈处理
                            this.codelanguage = 'java';
                            this.technologyStack.emit(this.codelanguage);
                            if (classifier.id === 'ref' || classifier.id === 'remote') {
                                this.initMavenSourceList();
                            }
                        }
                        else {
                            this.codelanguage = 'dotnet';
                            this.technologyStack.emit(this.codelanguage);
                            if (classifier.id === 'ref' || classifier.id === 'remote') {
                                this.initNugetSourceList();
                            }
                        }
                    }));
                }
            }));
            // 处理模式
            this.metadataService.getProcessMode().subscribe((/**
             * @param {?} processMode
             * @return {?}
             */
            processMode => {
                this.processMode = processMode;
            }));
        }));
        this.subscriptions.push(subscription);
    }
    /**
     * @return {?}
     */
    clear() {
        this.data = [];
        this.selectionChanged(null, []);
        this.page = {
            pageIndex: this.initPageIndex,
            pageSize: this.initPageSize
        };
    }
    /**
     * @param {?} state
     * @return {?}
     */
    dataStateChange(state) {
        this.state = state;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.subscriptions && this.subscriptions.length > 0) {
            for (const item of this.subscriptions) {
                item.unsubscribe();
            }
        }
    }
    /**
     * @return {?}
     */
    initNugetSourceList() {
        if (this.sourceList.length === 1) {
            this.gspMetaService.getNugetSourceList().subscribe((/**
             * @param {?} sources
             * @return {?}
             */
            sources => {
                sources.forEach((/**
                 * @param {?} source
                 * @return {?}
                 */
                source => {
                    this.sourceList.push({ text: source.name, value: source.name });
                }));
            }));
        }
    }
    /**
     * @return {?}
     */
    initMavenSourceList() {
        if (this.sourceList.length === 1) {
            this.gspMetaService.getMavenSourceList().subscribe((/**
             * @param {?} sources
             * @return {?}
             */
            sources => {
                if (sources.repoList) {
                    sources.repoList.forEach((/**
                     * @param {?} source
                     * @return {?}
                     */
                    source => {
                        this.sourceList.push({ text: source.id, value: source.id });
                    }));
                }
            }));
        }
    }
    /**
     * @return {?}
     */
    onSearchTextChanged() {
        if (!this.searchText) {
            this.searchText = '';
        }
        /** @type {?} */
        const searchTextTrim = this.searchText.trim();
        if (this.searchTextForDup === searchTextTrim) {
            return;
        }
        this.searchTextForDup = searchTextTrim;
        console.log(this.searchText);
        this.searchTextSub.next(searchTextTrim);
    }
    /**
     * @private
     * @return {?}
     */
    GetMetadataList() {
        /** @type {?} */
        const metadatadto = [];
        this.loading = true;
        this.metadataService.GetMetadataList(this.postfixSelected).subscribe((/**
         * @param {?} templates
         * @return {?}
         */
        templates => {
            this.loading = false;
            if (!templates) {
                this.gridData = this.data;
                gsp.ide.messager.error('error to get response');
                return;
            }
            // tslint:disable-next-line: prefer-for-of
            for (let index = 0; index < templates.length; index++) {
                /** @type {?} */
                const template = templates[index];
                if (this.searchText) {
                    // tslint:disable-next-line: max-line-length
                    if (template.code.indexOf(this.searchText) > -1 || template.name.indexOf(this.searchText) > -1) {
                        metadatadto.push(template);
                    }
                }
                else {
                    metadatadto.push(template);
                }
            }
            if (metadatadto.length > 0) {
                this.onSelectionChangedInternal(metadatadto[0], [0]);
            }
            this.hasNext = false;
            this.data = metadatadto;
            this.gridData = this.data;
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            this.data = metadatadto;
            this.gridData = this.data;
            this.loading = false;
            /** @type {?} */
            const msg = error.error.Message !== undefined ? error.error.Message : error.error.error.message;
            gsp.ide.messager.error('Get Metadatas Error:' + msg);
        }));
    }
    /**
     * @private
     * @return {?}
     */
    GetMetadataListInBo() {
        /** @type {?} */
        const metadatadto = [];
        this.loading = true;
        this.metadataService.GetMetadataListInBo(this.postfixSelected).subscribe((/**
         * @param {?} templates
         * @return {?}
         */
        templates => {
            this.loading = false;
            if (!templates) {
                this.gridData = this.data;
                gsp.ide.messager.error('error to get response');
                return;
            }
            // tslint:disable-next-line: prefer-for-of
            for (let index = 0; index < templates.length; index++) {
                /** @type {?} */
                const template = templates[index];
                if (this.searchText) {
                    // tslint:disable-next-line: max-line-length
                    if (template.projectName.toLowerCase().indexOf(this.searchText.toLowerCase()) > -1
                        || template.code.toLowerCase().indexOf(this.searchText.toLowerCase()) > -1
                        || template.name.toLowerCase().indexOf(this.searchText.toLowerCase()) > -1) {
                        metadatadto.push(template);
                    }
                }
                else {
                    metadatadto.push(template);
                }
            }
            if (metadatadto.length > 0) {
                this.onSelectionChangedInternal(metadatadto[0], [0]);
            }
            this.hasNext = false;
            this.data = metadatadto;
            this.gridData = this.data;
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            this.data = metadatadto;
            this.gridData = this.data;
            this.loading = false;
            /** @type {?} */
            const msg = error.error.Message !== undefined ? error.error.Message : error.error.error.message;
            gsp.ide.messager.error('Get Metadatas Error:' + msg);
        }));
    }
    /**
     * @private
     * @param {?=} reloadT
     * @return {?}
     */
    GetLocalMetadataList(reloadT = true) {
        /** @type {?} */
        const metadatadto = [];
        this.loading = true;
        // this.gspMetaService.GetDevRootPath().subscribe(root => {
        // this.metadataService.rootPath = root;
        // tslint:disable-next-line:max-line-length
        this.metadataService.GetMetadataInfos(this.scope, this.codelanguage, this.isMavenEnable, this.postfixSelected, this.searchText, this.page).subscribe((/**
         * @param {?} templates
         * @return {?}
         */
        (templates) => {
            this.loading = false;
            if (!templates) {
                gsp.ide.messager.error('error to get metadata infos');
                return;
            }
            // 页码处理
            if (templates.page) {
                if (templates.page.pageSize < this.initPageSize) {
                    this.hasNext = false;
                }
                else {
                    this.hasNext = true;
                }
                this.page.pageSize = templates.page.pageSize;
                this.page.pageIndex = templates.page.pageIndex;
            }
            else {
                this.hasNext = false;
            }
            /** @type {?} */
            const metadataList = templates.metadataIndexItems;
            if (!metadataList || metadataList.length < 1) {
                this.mySelection = [];
                this.gridData = this.data;
                return;
            }
            metadataList.forEach((/**
             * @param {?} element
             * @return {?}
             */
            element => {
                if (element.repoPkg && element.repoPkg.length > 0) {
                    element.nugetPkgName = element.repoPkg[0].name;
                    element.nugetPkgVersion = element.repoPkg[0].version;
                }
                metadatadto.push(element);
            }));
            if (reloadT) {
                this.onSelectionChangedInternal(metadatadto[0], [0]);
                this.data = metadatadto;
            }
            else {
                metadatadto.forEach((/**
                 * @param {?} dto
                 * @return {?}
                 */
                dto => {
                    this.data.push(dto);
                }));
            }
            this.gridData = this.data;
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            this.data = metadatadto;
            this.gridData = this.data;
            this.loading = false;
            /** @type {?} */
            const msg = error.error.Message !== undefined ? error.error.Message : error.error.error.message;
            gsp.ide.messager.error('Metadata req failed: ' + msg);
        }));
    }
    /**
     * @param {?} dataItem
     * @param {?} index
     * @return {?}
     */
    showOnlyBeveragesDetails(dataItem, index) {
        return dataItem.repoPkg && dataItem.repoPkg.length > 1;
    }
    /**
     * @param {?} reload
     * @return {?}
     */
    getRemoteMetadataList(reload) {
        /** @type {?} */
        const metadatadto = [];
        this.loading = true;
        // tslint:disable-next-line:max-line-length
        this.metadataService.GetMetadataInfos(this.scope, this.codelanguage, this.isMavenEnable, this.postfixSelected, this.searchText, this.page, this.selectedSource.value).subscribe((/**
         * @param {?} templates
         * @return {?}
         */
        (templates) => {
            this.loading = false;
            if (!templates) {
                gsp.ide.messager.error('error to get metadata infos');
                return;
            }
            // 页码处理
            if (templates.page) {
                if (templates.page.pageSize < this.initPageSize) {
                    this.hasNext = false;
                }
                else {
                    this.hasNext = true;
                }
                this.page.pageSize = templates.page.pageSize;
                this.page.pageIndex = templates.page.pageIndex;
            }
            else {
                this.hasNext = false;
            }
            /** @type {?} */
            const metadataList = templates.metadataIndexItems;
            if (!metadataList || metadataList.length < 1) {
                this.mySelection = [];
                this.gridData = this.data;
                return;
            }
            metadataList.forEach((/**
             * @param {?} element
             * @return {?}
             */
            element => {
                if (!element.repoPkg || element.repoPkg.length < 1) {
                    return;
                }
                if (this.codelanguage === 'dotnet' || this.isMavenEnable === false) {
                    element.repoPkg.sort(this.compareNuget('version'));
                    element.nugetPkgName = element.repoPkg[0].name;
                    element.nugetPkgVersion = element.repoPkg[0].version;
                }
                else {
                    element.repoPkg.sort(this.compare('version'));
                    element.groupId = element.repoPkg[0].groupId;
                    element.artifactId = element.repoPkg[0].artifactId;
                    element.mavenPkgVersion = element.repoPkg[0].version;
                }
                metadatadto.push(element);
            }));
            if (!metadatadto || metadatadto.length < 1) {
                this.mySelection = [];
                this.gridData = this.data;
                return;
            }
            if (reload) {
                this.onSelectionChangedInternal(metadatadto[0], [0]);
                this.data = metadatadto;
            }
            else {
                metadatadto.forEach((/**
                 * @param {?} dto
                 * @return {?}
                 */
                dto => {
                    this.data.push(dto);
                }));
            }
            this.gridData = this.data;
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            this.data = metadatadto;
            this.gridData = this.data;
            this.loading = false;
            /** @type {?} */
            const msg = error.error.Message !== undefined ? error.error.Message : error.error.error.message;
            gsp.ide.messager.error('Metadata req failed: ' + msg);
        }));
    }
    /**
     * @param {?} pro
     * @return {?}
     */
    compare(pro) {
        return (/**
         * @param {?} obj1
         * @param {?} obj2
         * @return {?}
         */
        (obj1, obj2) => {
            /** @type {?} */
            const val1 = obj1[pro];
            /** @type {?} */
            const val2 = obj2[pro];
            /** @type {?} */
            const vala1 = val1.split('.');
            /** @type {?} */
            const vala2 = val2.split('.');
            /** @type {?} */
            const v10 = vala1[0].substr(1);
            /** @type {?} */
            const v20 = vala2[0].substr(1);
            if (Number(v10) < Number(v20)) {
                return 1;
            }
            if (Number(v10) > Number(v20)) {
                return -1;
            }
            if (Number(vala1[1]) < Number(vala2[1])) {
                return 1;
            }
            if (Number(vala1[1]) > Number(vala2[1])) {
                return -1;
            }
            /** @type {?} */
            const v12 = vala1[2].replace('-SNAPSHOT', '');
            /** @type {?} */
            const v22 = vala2[2].replace('-SNAPSHOT', '');
            if (Number(v12) < Number(v22)) {
                return 1;
            }
            if (Number(v12) > Number(v22)) {
                return -1;
            }
            if (vala1[2].indexOf('-SNAPSHOT') > 0) {
                return 1;
            }
            else {
                return -1;
            }
        });
    }
    /**
     * @param {?} pro
     * @return {?}
     */
    compareNuget(pro) {
        return (/**
         * @param {?} obj1
         * @param {?} obj2
         * @return {?}
         */
        (obj1, obj2) => {
            /** @type {?} */
            const val1 = obj1[pro];
            /** @type {?} */
            const val2 = obj2[pro];
            /** @type {?} */
            const vala1 = val1.split('.');
            /** @type {?} */
            const vala2 = val2.split('.');
            /** @type {?} */
            const v10 = vala1[0];
            /** @type {?} */
            const v20 = vala2[0];
            if (Number(v10) < Number(v20)) {
                return -1;
            }
            if (Number(v10) > Number(v20)) {
                return 1;
            }
            if (Number(vala1[1]) < Number(vala2[1])) {
                return 1;
            }
            if (Number(vala1[1]) > Number(vala2[1])) {
                return -1;
            }
            /** @type {?} */
            const v12 = vala1[2];
            /** @type {?} */
            const v22 = vala2[2];
            if (v12.indexOf('alpha') > 0) {
                if (v12 < v22) {
                    return 1;
                }
                if (v12 > v22) {
                    return -1;
                }
            }
            else {
                if (Number(v12) < Number(v22)) {
                    return 1;
                }
                if (Number(v12) > Number(v22)) {
                    return -1;
                }
            }
        });
    }
    /**
     * @param {?} classifier
     * @return {?}
     */
    scopeChanged(classifier) {
        this.classifier = classifier;
        this.getData();
    }
    /**
     * @return {?}
     */
    getData() {
        if (this.classifier) {
            if (this.reload) {
                this.clear();
            }
            /** @type {?} */
            const reloadT = this.reload;
            if (this.classifier.id === 'project') {
                this.scope = 0;
                this.GetMetadataList();
            }
            else if (this.classifier.id === 'bo') {
                this.scope = 0;
                this.GetMetadataListInBo();
            }
            else if (this.classifier.id === 'ref') {
                this.scope = 3;
                this.getRemoteMetadataList(reloadT);
            }
            else if (this.classifier.id === 'remote') {
                this.scope = 2;
                this.getRemoteMetadataList(reloadT);
            }
            else if (this.classifier.id === 'local') {
                this.scope = 1;
                this.GetLocalMetadataList(reloadT);
            }
            // 重置一次请求中的状态
            this.reload = true;
        }
    }
    // 选中元数据后的事件处理
    /**
     * @param {?} event
     * @return {?}
     */
    onSelectionChange(event) {
        if (event && event.selectedRows && event.selectedRows.length > 0) {
            /** @type {?} */
            const dataItem = event.selectedRows[0].dataItem;
            this.onSelectionChangedInternal(dataItem, [event.selectedRows[0].index]);
        }
    }
    /**
     * @private
     * @param {?} dataItem
     * @param {?} index
     * @return {?}
     */
    onSelectionChangedInternal(dataItem, index) {
        /** @type {?} */
        const newData = this.makeNewData(dataItem);
        if (this.scope === 2 || this.scope === 3) {
            newData.repoPkg.push(dataItem.repoPkg[0]);
        }
        this.selectionChanged(newData, index);
    }
    /**
     * @private
     * @param {?} dataItem
     * @param {?=} index
     * @return {?}
     */
    selectionChanged(dataItem, index = []) {
        this.selectedMetadataChanged.emit(dataItem);
        this.mySelection = index;
    }
    /**
     * @private
     * @param {?} dataItem
     * @return {?}
     */
    makeNewData(dataItem) {
        return {
            mdPkgName: dataItem.mdPkgName,
            repoPkg: [],
            id: dataItem.id,
            nameSpace: dataItem.nameSpace,
            code: dataItem.code,
            name: dataItem.name,
            fileName: dataItem.fileName,
            type: dataItem.type,
            bizobjectID: dataItem.bizobjectID,
            language: dataItem.language,
            isTranslating: dataItem.isTranslating,
            relativePath: dataItem.relativePath,
            extendProperty: dataItem.extendProperty,
            processMode: dataItem.processMode
        };
    }
    /**
     * @param {?} dataItem
     * @return {?}
     */
    onSelectedPkgChanged(dataItem) {
        if (dataItem && dataItem.repoPkg && dataItem.repoPkg.length > 0) {
            /** @type {?} */
            const newData = this.makeNewData(dataItem);
            if (dataItem.selectedPkg) {
                newData.repoPkg.push(dataItem.selectedPkg);
            }
            else {
                newData.repoPkg.push(dataItem.repoPkg[0]);
            }
            this.selectionChanged(newData, []);
        }
    }
    /**
     * @return {?}
     */
    goSearch() {
        if (!this.searchText) {
            this.searchText = '';
        }
        console.log('搜索');
        /** @type {?} */
        const searchTextTrim = this.searchText.trim();
        if (this.searchText && this.searchTextForDup === searchTextTrim) {
            return;
        }
        this.searchTextForDup = searchTextTrim;
        console.log(this.searchText);
        this.searchText = searchTextTrim;
        // 搜索
        console.log('search for ' + this.searchText);
        this.getData();
    }
    /**
     * @return {?}
     */
    goNext() {
        console.log('下一页');
        this.page.pageIndex = this.page.pageIndex + 1;
        this.reload = false;
        this.getData();
    }
    /**
     * @return {?}
     */
    onConflictAvoidFlagChanged() {
        console.log('ConflictAvoidFlag ' + this.conflictAvoidFlag);
        this.getData();
    }
    /**
     * @param {?} mdSeleted
     * @return {?}
     */
    typeSelectedChanged(mdSeleted) {
        if (!mdSeleted) {
            mdSeleted = this.selectedType;
        }
        if (mdSeleted) {
            if (mdSeleted.value === this.allType.value) {
                this.postfixSelected = this.postfix;
            }
            else {
                this.postfixSelected = mdSeleted.value;
            }
            this.getData();
        }
    }
    /**
     * @param {?} source
     * @return {?}
     */
    sourceSelectedChanged(source) {
        if (!source) {
            source = this.selectedSource;
        }
        this.selectedSource = source;
        this.getData();
    }
}
TemplatemetadataComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'app-metadatalist',
                template: "<div class=\"templateInfo farris-main-area flex-column\">\r\n  <div class=\"templateSelectorHeader m-2\">\r\n    <div class=\"searchArea d-flex flex-row\">\r\n\r\n      <input id=\"url\" class=\"textInput flex-fill\" type=\"text\" size=\"30\" kendoTextBox [(ngModel)]=\"this.searchText\"\r\n        #url=\"ngModel\" required (ngModelChange)=\"onSearchTextChanged()\">\r\n      <button type=\"button\" class=\"btn btn-light btn-sm px-3 mx-2\"\r\n          (click)=\"goSearch()\">\u641C\u7D22</button>\r\n\r\n      <!-- <div class=\"conflictAvoidFlag\" *ngIf=\"this.scope === 2\">\r\n        <input type=\"checkbox\" class=\"form-check-input\" id=\"check1\"\r\n          [(ngModel)]=\"this.conflictAvoidFlag\" (ngModelChange)=\"this.onConflictAvoidFlagChanged()\">\r\n        <label class=\"form-check-label text-label\" for=\"check1\">\u907F\u514D\u51B2\u7A81</label>\r\n      </div> -->\r\n\r\n      <kendo-combobox class=\"mx-2\" [allowCustom]=\"allowCustom\" *ngIf=\"this.scope === 2 || this.scope === 3\" [data]=\"sourceList\" \r\n        [textField]=\"'text'\" [valueField]=\"'value'\" [(ngModel)]=\"this.selectedSource\" \r\n        (selectionChange)=\"sourceSelectedChanged($event)\" [popupSettings]=\"popupSettings\">\r\n      </kendo-combobox>\r\n\r\n      <kendo-combobox [readonly]=\"this.typeListReadonly\" [data]=\"typeList\" [(ngModel)]=\"selectedType\"\r\n        [textField]=\"'text'\" [valueField]=\"'value'\"  (selectionChange)=\"typeSelectedChanged($event)\">\r\n      </kendo-combobox>\r\n    </div>\r\n  </div>\r\n  <div class=\"templateSelector flex-fill d-flex flex-column\">\r\n    <kendo-grid class=\"flex-fill\"\r\n                [data]=\"gridData\"\r\n                [selectable]=\"selectableSetting\" \r\n                (selectionChange)=onSelectionChange($event)\r\n                [sortable]=\"true\"\r\n                [pageable]=\"false\"\r\n                [filterable]=\"false\"\r\n                [resizable]=\"true\"\r\n                [loading]=\"this.loading\"\r\n                [rowSelected]=\"rowSelected\">\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"(this.scope === 0 && this.classifier.id === 'bo') && (this.isMavenEnable === true && this.codelanguage === java)\" field=\"projectName\" title=\"\u5DE5\u7A0B\u540D\u79F0\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"150\" field=\"code\" title=\"\u7F16\u53F7\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"150\" field=\"name\" title=\"\u540D\u79F0\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column field=\"type\" [width]=\"130\" title=\"\u7C7B\u578B\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"this.scope === 1 || this.scope === 2 || this.scope === 3\" field=\"mdPkgName\"\r\n        title=\"\u5143\u6570\u636E\u5305\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"(this.scope === 2 || this.scope === 3) && (this.isMavenEnable === false || this.codelanguage === dotnet)\" field=\"nugetPkgName\" title=\"\u6240\u5C5ENuget\u5305\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"(this.scope === 2 || this.scope === 3) && (this.isMavenEnable === false || this.codelanguage === dotnet)\" field=\"nugetPkgVersion\" title=\"\u6240\u5C5ENuget\u5305\u7248\u672C\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"(this.scope === 2 || this.scope === 3) && (this.isMavenEnable === true && this.codelanguage === java)\" field=\"groupId\" title=\"\u6240\u5C5EMaven\u5305GroupId\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"(this.scope === 2 || this.scope === 3) && (this.isMavenEnable === true && this.codelanguage === java)\" field=\"artifactId\" title=\"\u6240\u5C5EMaven\u5305ArtifactId\">\r\n      </kendo-grid-column>\r\n      <kendo-grid-column [width]=\"190\" *ngIf=\"(this.scope === 2 || this.scope === 3) && (this.isMavenEnable === true && this.codelanguage === java)\" field=\"mavenPkgVersion\" title=\"\u6240\u5C5EMaven\u5305\u7248\u672C\">\r\n      </kendo-grid-column>\r\n\r\n      <ng-template kendoGridDetailTemplate let-dataItem [kendoGridDetailTemplateShowIf]=\"showOnlyBeveragesDetails\">\r\n        <lib-metadatalistdetails (selectedPkgChanged)=\"this.onSelectedPkgChanged($event)\" [isMavenEnable]=\"isMavenEnable\" [mdItem]=\"dataItem\" [codelanguage]=\"codelanguage\">\r\n        </lib-metadatalistdetails>\r\n      </ng-template>\r\n\r\n    </kendo-grid>\r\n  </div>\r\n  <div *ngIf=\"this.hasNext && !this.loading\" class=\"model-footer\">\r\n    <button type=\"button\" class=\"btn btn-link btn-sm\" (click)=\"goNext()\">\u52A0\u8F7D\u66F4\u591A</button>\r\n  </div>\r\n  <div *ngIf=\"this.loading\" class=\"model-footer\">\r\n    <label class=\"text-label\">\u52A0\u8F7D\u4E2D...</label>\r\n  </div>\r\n  <div *ngIf=\"!this.hasNext && !this.loading\" class=\"model-footer\">\r\n    <label class=\"text-label\">\u5DF2\u7ECF\u5230\u5E95\u4E86</label>\r\n  </div>\r\n</div>",
                styles: [".templateSelector{overflow-y:hidden;overflow-x:auto}.templateInfo{display:flex;overflow:hidden}.conflictAvoidFlag{margin-left:25px;align-content:baseline;margin-top:1px;margin-right:3px;min-width:65px}.scope1{min-width:800px}.scope2{min-width:1000px}.text-label{margin-top:1.3px}"]
            }] }
];
/** @nocollapse */
TemplatemetadataComponent.ctorParameters = () => [
    { type: MetadataService },
    { type: GSPMetadataService }
];
TemplatemetadataComponent.propDecorators = {
    isMavenEnable: [{ type: Input }],
    technologyStack: [{ type: Output }],
    typeName: [{ type: Input }],
    filter: [{ type: Input }],
    selectedMetadataChanged: [{ type: Output }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    TemplatemetadataComponent.prototype.subscriptions;
    /** @type {?} */
    TemplatemetadataComponent.prototype.classifier;
    /** @type {?} */
    TemplatemetadataComponent.prototype.codelanguage;
    /** @type {?} */
    TemplatemetadataComponent.prototype.processMode;
    /** @type {?} */
    TemplatemetadataComponent.prototype.isMavenEnable;
    /** @type {?} */
    TemplatemetadataComponent.prototype.dotnet;
    /** @type {?} */
    TemplatemetadataComponent.prototype.java;
    /** @type {?} */
    TemplatemetadataComponent.prototype.technologyStack;
    /** @type {?} */
    TemplatemetadataComponent.prototype.loading;
    /** @type {?} */
    TemplatemetadataComponent.prototype.typeName;
    /** @type {?} */
    TemplatemetadataComponent.prototype.filter;
    /** @type {?} */
    TemplatemetadataComponent.prototype.selectedMetadataChanged;
    /** @type {?} */
    TemplatemetadataComponent.prototype.allSource;
    /** @type {?} */
    TemplatemetadataComponent.prototype.selectedSource;
    /** @type {?} */
    TemplatemetadataComponent.prototype.sourceList;
    /** @type {?} */
    TemplatemetadataComponent.prototype.popupSettings;
    /** @type {?} */
    TemplatemetadataComponent.prototype.allowCustom;
    /** @type {?} */
    TemplatemetadataComponent.prototype.allType;
    /** @type {?} */
    TemplatemetadataComponent.prototype.selectedType;
    /** @type {?} */
    TemplatemetadataComponent.prototype.typeList;
    /** @type {?} */
    TemplatemetadataComponent.prototype.postfixSelected;
    /** @type {?} */
    TemplatemetadataComponent.prototype.initPageIndex;
    /** @type {?} */
    TemplatemetadataComponent.prototype.initPageSize;
    /** @type {?} */
    TemplatemetadataComponent.prototype.page;
    /** @type {?} */
    TemplatemetadataComponent.prototype.typeListReadonly;
    /** @type {?} */
    TemplatemetadataComponent.prototype.hasNext;
    /** @type {?} */
    TemplatemetadataComponent.prototype.searchText;
    /** @type {?} */
    TemplatemetadataComponent.prototype.searchTextForDup;
    /** @type {?} */
    TemplatemetadataComponent.prototype.searchTextSub;
    /** @type {?} */
    TemplatemetadataComponent.prototype.postfix;
    /** @type {?} */
    TemplatemetadataComponent.prototype.postfixFlag;
    /** @type {?} */
    TemplatemetadataComponent.prototype.scope;
    /** @type {?} */
    TemplatemetadataComponent.prototype.data;
    /** @type {?} */
    TemplatemetadataComponent.prototype.conflictAvoidFlag;
    /** @type {?} */
    TemplatemetadataComponent.prototype.reload;
    /** @type {?} */
    TemplatemetadataComponent.prototype.mySelection;
    /** @type {?} */
    TemplatemetadataComponent.prototype.state;
    /** @type {?} */
    TemplatemetadataComponent.prototype.gridData;
    /** @type {?} */
    TemplatemetadataComponent.prototype.selectableSetting;
    /** @type {?} */
    TemplatemetadataComponent.prototype.rowSelected;
    /**
     * @type {?}
     * @private
     */
    TemplatemetadataComponent.prototype.metadataService;
    /**
     * @type {?}
     * @private
     */
    TemplatemetadataComponent.prototype.gspMetaService;
}
