// @edp-pmf/bpmn-model/dist/esm/lib/model/mex/ExecutionListener.js
import { __extends } from "tslib";
import { BpmnModelHelper } from "../BpmnModelHelper";
import { MexExtensionElement } from "./MexExtensionElement";
var ExecutionListener = /** @class */ (function (_super) {
    __extends(ExecutionListener, _super);
    function ExecutionListener(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.Mex.ExecutionListener";
        return _this;
    }
    ExecutionListener.prototype.NewInstance = function () {
        return new ExecutionListener(this.Model);
    };
    ExecutionListener.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Class = null;
        this.Event = null;
        ;
    };
    ExecutionListener.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "Event", this.Event);
        BpmnModelHelper.SetAttributeValue(obj, "Class", this.Class);
        return obj;
    };
    ExecutionListener.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.Event = obj.Event;
        this.Class = obj.Class;
    };
    return ExecutionListener;
}(MexExtensionElement));
export { ExecutionListener };
