// @edp-pmf/bpmn-model/dist/esm/lib/model/mex/PalUserTaskAttribute.js
import { __extends } from "tslib";
import { BpmnModelJsonHelper } from "../BpmnModelJsonHelper";
import { ExtensionMethod } from "../ExtensionMethod";
import { PalParticipant } from "../participant";
import { MexElement } from "./MexElement";
var PalUserTaskAttribute = /** @class */ (function (_super) {
    __extends(PalUserTaskAttribute, _super);
    function PalUserTaskAttribute(model) {
        var _this = _super.call(this, model) || this;
        _this.responsiblePersons = [];
        _this.accountablePersons = [];
        _this.consultedPersons = [];
        _this.informedPersons = [];
        _this.formDataIds = [];
        _this.itSystemIds = [];
        _this.riskIds = [];
        _this.riskManagementIds = [];
        _this.systemIds = [];
        _this.kpiIds = [];
        return _this;
    }
    PalUserTaskAttribute.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.name = null;
        this.layer = 0;
        this.description = null;
        this.code = null;
        this.input = null;
        this.output = null;
        this.responsiblePersons = [];
        this.accountablePersons = [];
        this.consultedPersons = [];
        this.informedPersons = [];
        this.formDataIds = [];
        this.itSystemIds = [];
        this.riskIds = [];
        this.riskManagementIds = [];
        this.systemIds = [];
        this.kpiIds = [];
    };
    PalUserTaskAttribute.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        ExtensionMethod.SetAttributeValue(obj, "name", this.name);
        ExtensionMethod.SetAttributeValue(obj, "layer", this.layer);
        ExtensionMethod.SetAttributeValue(obj, "description", this.description);
        ExtensionMethod.SetAttributeValue(obj, "code", this.code);
        ExtensionMethod.SetAttributeValue(obj, "input", this.input);
        ExtensionMethod.SetAttributeValue(obj, "output", this.output);
        ExtensionMethod.SetAttributeValue(obj, "responsiblePersons", this.responsiblePersons ? this.responsiblePersons.map(function (it) { return it.ConvertToJsonObject(); }) : []);
        ExtensionMethod.SetAttributeValue(obj, "accountablePersons", this.accountablePersons ? this.accountablePersons.map(function (it) { return it.ConvertToJsonObject(); }) : []);
        ExtensionMethod.SetAttributeValue(obj, "consultedPersons", this.consultedPersons ? this.consultedPersons.map(function (it) { return it.ConvertToJsonObject(); }) : []);
        ExtensionMethod.SetAttributeValue(obj, "informedPersons", this.informedPersons ? this.informedPersons.map(function (it) { return it.ConvertToJsonObject(); }) : []);
        ExtensionMethod.SetAttributeValue(obj, "formDataIds", BpmnModelJsonHelper.StringListToJson(this.formDataIds));
        ExtensionMethod.SetAttributeValue(obj, "itSystemIds", BpmnModelJsonHelper.StringListToJson(this.itSystemIds));
        ExtensionMethod.SetAttributeValue(obj, "riskIds", BpmnModelJsonHelper.StringListToJson(this.riskIds));
        ExtensionMethod.SetAttributeValue(obj, "riskManagementIds", BpmnModelJsonHelper.StringListToJson(this.riskManagementIds));
        ExtensionMethod.SetAttributeValue(obj, "systemIds", BpmnModelJsonHelper.StringListToJson(this.systemIds));
        ExtensionMethod.SetAttributeValue(obj, "kpiIds", BpmnModelJsonHelper.StringListToJson(this.kpiIds));
        return obj;
    };
    PalUserTaskAttribute.prototype.LoadFromJsonObject = function (obj) {
        var _this = this;
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.name = ExtensionMethod.GetAttributeValue4String(obj, "name");
        this.layer = ExtensionMethod.GetAttributeValue4Int64(obj, "layer");
        this.description = ExtensionMethod.GetAttributeValue4String(obj, "description");
        this.code = ExtensionMethod.GetAttributeValue4String(obj, "code");
        this.input = ExtensionMethod.GetAttributeValue4String(obj, "input");
        this.output = ExtensionMethod.GetAttributeValue4String(obj, "output");
        var responsiblePersonObjs = ExtensionMethod.GetAttributeValue(obj, "responsiblePersons");
        this.responsiblePersons = responsiblePersonObjs ? responsiblePersonObjs.map(function (it) {
            var d = new PalParticipant(_this.Model);
            d.LoadFromJsonObject(it);
            return d;
        }) : [];
        var accountablePersonObjs = ExtensionMethod.GetAttributeValue(obj, "accountablePersons");
        this.accountablePersons = accountablePersonObjs ? accountablePersonObjs.map(function (it) {
            var d = new PalParticipant(_this.Model);
            d.LoadFromJsonObject(it);
            return d;
        }) : [];
        var consultedPersonObjs = ExtensionMethod.GetAttributeValue(obj, "consultedPersons");
        this.consultedPersons = consultedPersonObjs ? consultedPersonObjs.map(function (it) {
            var d = new PalParticipant(_this.Model);
            d.LoadFromJsonObject(it);
            return d;
        }) : [];
        var informedPersonObjs = ExtensionMethod.GetAttributeValue(obj, "informedPersons");
        this.informedPersons = informedPersonObjs ? informedPersonObjs.map(function (it) {
            var d = new PalParticipant(_this.Model);
            d.LoadFromJsonObject(it);
            return d;
        }) : [];
        this.formDataIds = BpmnModelJsonHelper.StringListFromJson(ExtensionMethod.GetAttributeValue4String(obj, "formDataIds"));
        this.itSystemIds = BpmnModelJsonHelper.StringListFromJson(ExtensionMethod.GetAttributeValue4String(obj, "itSystemIds"));
        this.riskIds = BpmnModelJsonHelper.StringListFromJson(ExtensionMethod.GetAttributeValue4String(obj, "riskIds"));
        this.riskManagementIds = BpmnModelJsonHelper.StringListFromJson(ExtensionMethod.GetAttributeValue4String(obj, "riskManagementIds"));
        this.systemIds = BpmnModelJsonHelper.StringListFromJson(ExtensionMethod.GetAttributeValue4String(obj, "systemIds"));
        this.kpiIds = BpmnModelJsonHelper.StringListFromJson(ExtensionMethod.GetAttributeValue4String(obj, "kpiIds"));
    };
    PalUserTaskAttribute.prototype.clone = function () {
        var c = new PalUserTaskAttribute(this.Model);
        c.name = this.name;
        c.layer = this.layer;
        c.description = this.description;
        c.code = this.code;
        c.input = this.input;
        c.output = this.output;
        c.responsiblePersons = this.responsiblePersons.map(function (it) { return it.clone(); });
        c.accountablePersons = this.accountablePersons.map(function (it) { return it.clone(); });
        c.consultedPersons = this.consultedPersons.map(function (it) { return it.clone(); });
        c.informedPersons = this.informedPersons.map(function (it) { return it.clone(); });
        c.formDataIds = this.formDataIds ? this.formDataIds.map(function (it) { return it; }) : [];
        c.itSystemIds = this.itSystemIds ? this.itSystemIds.map(function (it) { return it; }) : [];
        c.riskIds = this.riskIds ? this.riskIds.map(function (it) { return it; }) : [];
        c.riskManagementIds = this.riskManagementIds ? this.riskManagementIds.map(function (it) { return it; }) : [];
        c.systemIds = this.systemIds ? this.systemIds.map(function (it) { return it; }) : [];
        c.kpiIds = this.kpiIds ? this.kpiIds.map(function (it) { return it; }) : [];
        return c;
    };
    return PalUserTaskAttribute;
}(MexElement));
export { PalUserTaskAttribute };
