// @edp-pmf/bpmn-model/dist/esm/lib/model/mex/VariableExpression.js
import { __extends } from "tslib";
import { BpmnModelHelper } from "../BpmnModelHelper";
import { MexExpression } from "./MexExpression";
import { VariableExpressionDirection } from "./VariableExpressionDirection";
var VariableExpression = /** @class */ (function (_super) {
    __extends(VariableExpression, _super);
    function VariableExpression(model) {
        var _this = _super.call(this, model) || this;
        _this.ClrTypeID = "bpmn.Mex.VariableExpression";
        _this.Direction = VariableExpressionDirection.ActivityDataInputAssociation;
        return _this;
    }
    VariableExpression.prototype.NewInstance = function () {
        return new VariableExpression(this.Model);
    };
    VariableExpression.prototype.ResetBeforeLoadFrom = function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.Direction = null;
        this.SourceRef = null;
    };
    VariableExpression.prototype.ConvertToJsonObject = function () {
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelHelper.SetAttributeValue(obj, "SourceRef", this.SourceRef);
        BpmnModelHelper.SetAttributeValue(obj, "Direction", this.Direction);
        return obj;
    };
    VariableExpression.prototype.LoadFromJsonObject = function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.SourceRef = obj.SourceRef;
        this.Direction = obj.Direction;
    };
    return VariableExpression;
}(MexExpression));
export { VariableExpression };
