// @gspwidget/util/esm2015/lib/repositories/page.repo.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map, tap } from 'rxjs/operators';
import { addBaseUrl } from '../const';
import { mapEntityFeilds } from '../utils/field-mapping';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
/** @type {?} */
export const pageInfoUrls = {
    GET_BY_ID_MANAGE: `page/`,
    GET_BY_ID_PERSONAL: `page/pageofuser?pageId=`,
    GET_ALL_MANAGE: `page/all/mg`,
    GET_ALL_PERSONAL: `page/pagelistofuser`,
    GET_BY_ROLE: `pagerole/pagesbyrole?roleId=`,
    GET_PAGES_WITH_ROLE: `pagerole/all`,
    SET_TO_ROLE_DEFAULT: `pagerole/settingrolepageenable`,
    // roleId pageId
    DELETE: `page/`,
    // delete + id
    DELETE_MANAGE: `page/delete/mg`,
    // post []
    ADD: `page`,
    ADD_MANAGE: `page/add/mg`,
    UPDATE: `page`,
    UPDATE_MANAGE: `page/update/mg`,
    COPY: `page/copy`,
    COPY_MANAGE: `page/copy/mg`,
    RESET_PAGE: `page/resetsyspage?pageId=`,
    SELECT_PAGE: `page/changepageselect?pageId=`,
};
/** @type {?} */
export const pageRoles = {
    SAVE: `pagerole/save?pageId=`,
    GET_BY_PAGE_ID: `pagerole/getallpagerole?pageId=`,
};
addBaseUrl(pageInfoUrls);
addBaseUrl(pageRoles);
export class PageRepo {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @return {?}
     */
    getPagesManage() {
        return this.http
            .get(pageInfoUrls.GET_ALL_MANAGE)
            .pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        data => data.map(convertPageInfo))));
    }
    /**
     * @return {?}
     */
    getPagesUser() {
        return this.http
            .get(pageInfoUrls.GET_ALL_PERSONAL);
    }
    /**
     * @param {?} roleId
     * @return {?}
     */
    getPagesByRole(roleId) {
        return this.http
            .get(pageInfoUrls.GET_BY_ROLE + roleId);
    }
    /**
     * @return {?}
     */
    getPagesWithRole() {
        return this.http.get(pageInfoUrls.GET_PAGES_WITH_ROLE);
    }
    /**
     * @param {?} roleId
     * @param {?} pageId
     * @return {?}
     */
    setRoleDefaultPage(roleId, pageId) {
        return this.http
            .post(pageInfoUrls.SET_TO_ROLE_DEFAULT, null, {
            params: { pageId, roleId }
        }).pipe(this.success());
    }
    /**
     * @param {?} id
     * @param {?=} isManage
     * @return {?}
     */
    getPageInfoById(id, isManage = false) {
        return this.http
            .get((isManage
            ? pageInfoUrls.GET_BY_ID_MANAGE
            : pageInfoUrls.GET_BY_ID_PERSONAL)
            + id)
            .pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        data => convertPageInfo(data))));
    }
    /**
     * @param {?} pageInfo
     * @param {?=} isManage
     * @return {?}
     */
    savePageInfo(pageInfo, isManage = false) {
        return pageInfo.id
            ? this.http.post(isManage ? pageInfoUrls.UPDATE_MANAGE : pageInfoUrls.UPDATE, convertPageInfoInverse(pageInfo)).pipe(this.success())
            : this.http.put(isManage ? pageInfoUrls.ADD_MANAGE : pageInfoUrls.ADD, convertPageInfoInverse(pageInfo)).pipe(this.success());
    }
    /**
     * @param {?} id
     * @return {?}
     */
    deletePageInfo(id) {
        return this.http
            .delete(pageInfoUrls.DELETE + id, { responseType: 'text' })
            .pipe(this.success());
    }
    /**
     * @param {?} ids
     * @return {?}
     */
    deletePageInfoManage(ids) {
        return this.http
            .post(pageInfoUrls.DELETE_MANAGE, ids)
            .pipe(this.success());
    }
    /**
     * @param {?} ids
     * @param {?=} isManage
     * @return {?}
     */
    copyPageInfo(ids, isManage = false) {
        return this.http
            .post(isManage ? pageInfoUrls.COPY_MANAGE : pageInfoUrls.COPY, ids, { responseType: 'text' });
    }
    /**
     * @param {?} pageId
     * @return {?}
     */
    resetPage(pageId) {
        return this.http.post(pageInfoUrls.RESET_PAGE + pageId, null);
    }
    /**
     * @param {?} pageId
     * @return {?}
     */
    selectPage(pageId) {
        return this.http.post(pageInfoUrls.SELECT_PAGE + pageId, null);
    }
    // TODO: 不该在这依赖 nzmsg，应移到使用处
    /**
     * @private
     * @return {?}
     */
    success() {
        return tap((/**
         * @return {?}
         */
        () => {
            // this.nzMsg.success(`操作成功！`, { nzDuration: 1000 })
        }));
    }
    /**
     * page roles
     * @param {?} pageId
     * @param {?} roleIds
     * @return {?}
     */
    savePageRoles(pageId, roleIds) {
        return this.http.post(pageRoles.SAVE + pageId, roleIds);
    }
    /**
     * @param {?} pageId
     * @return {?}
     */
    getPageRoles(pageId) {
        return this.http.get(pageRoles.GET_BY_PAGE_ID + pageId);
    }
}
PageRepo.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
PageRepo.ctorParameters = () => [
    { type: HttpClient }
];
/** @nocollapse */ PageRepo.ngInjectableDef = i0.defineInjectable({ factory: function PageRepo_Factory() { return new PageRepo(i0.inject(i1.HttpClient)); }, token: PageRepo, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    PageRepo.prototype.http;
}
/**
 * pageInfo 后端 -> 前端, 前端 -> 后端 转换
 * @param {?} original
 * @return {?}
 */
export function convertPageInfo(original) {
    /** @type {?} */
    const result = mapEntityFeilds(original, 'pageInfo');
    result.config = JSON.parse((/** @type {?} */ (result.config)));
    return result;
}
/**
 * @param {?} pageInfo
 * @return {?}
 */
export function convertPageInfoInverse(pageInfo) {
    /** @type {?} */
    const result = mapEntityFeilds(pageInfo, 'pageInfo', true)
    // 保存时，删除 pageInfo.config.layout 的元素的 portlet
    ;
    // 保存时，删除 pageInfo.config.layout 的元素的 portlet
    result.config.layout = result.config.layout.map((/**
     * @param {?} item
     * @return {?}
     */
    item => {
        const { instanceId, isBAPortlet, gridsterItem, portletId, link } = item;
        return { instanceId, isBAPortlet, gridsterItem, portletId, link };
    }));
    result.config = JSON.stringify(result.config);
    return result;
}
