// @gspwidget/util/esm2015/lib/repositories/portlet.repo.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { addBaseUrl } from '../const';
import { EppPortlet } from '../types/portlet.type';
import { convertWidget } from './widget.repo';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
/** @type {?} */
export const portletUrls = {
    GET_BY_ID: `portlet/`,
    GET_USER_ALL: `portlet/all`,
    SAVE: `portlet/`,
    GET_BY_MODULE_ID: `portlet/findByModuleId`,
    DELETE_BY_ID: `portlet/`,
    GET_PAGE: `page/findByPortletId`
};
/** @type {?} */
export const portletOps = {
    SAVE: `portletop/save?portletId=`,
    GET_BY_PORTLET_ID: `portletop/getallpagerole?portletId=`
};
addBaseUrl(portletUrls);
addBaseUrl(portletOps);
export class PortletRepo {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @param {?} portletId
     * @return {?}
     */
    getPageByPortletId(portletId) {
        return this.http.get(portletUrls.GET_PAGE, {
            params: { portletId }
        });
    }
    /**
     * @return {?}
     */
    getPortlets() {
        return this.http.get(portletUrls.GET_USER_ALL).pipe(map((/**
         * @param {?} portlets
         * @return {?}
         */
        portlets => portlets.map(convertPortlet))));
    }
    /**
     * @param {?=} moduleId
     * @return {?}
     */
    getPortletsByModuleId(moduleId) {
        return this.http.get(portletUrls.GET_BY_MODULE_ID, { params: moduleId ? { moduleId } : null }).pipe(map((/**
         * @param {?} portlets
         * @return {?}
         */
        portlets => portlets.map(convertPortlet))));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    getPortletById(id) {
        // return this.storage.getPortletById(id)
        return this.http.get(portletUrls.GET_BY_ID + id).pipe(map(convertPortlet));
    }
    /**
     * @param {?} portlet
     * @return {?}
     */
    savePortlet(portlet) {
        /** @type {?} */
        const serverPortlet = convertPortletInverse(portlet);
        return portlet.id
            ? this.http.post(portletUrls.SAVE, serverPortlet)
            : this.http.put(portletUrls.SAVE, serverPortlet);
    }
    /**
     * portlet op
     * @param {?} portletId
     * @return {?}
     */
    getPortletOps(portletId) {
        return this.http.get(portletOps.GET_BY_PORTLET_ID + portletId);
    }
    /**
     * @param {?} portletId
     * @param {?} opIds
     * @return {?}
     */
    savePortletOps(portletId, opIds) {
        return this.http.post(portletOps.SAVE + portletId, opIds);
    }
    /**
     * @param {?} portletId
     * @return {?}
     */
    deletePortletById(portletId) {
        return this.http.delete(portletUrls.DELETE_BY_ID + portletId);
    }
}
PortletRepo.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
PortletRepo.ctorParameters = () => [
    { type: HttpClient }
];
/** @nocollapse */ PortletRepo.ngInjectableDef = i0.defineInjectable({ factory: function PortletRepo_Factory() { return new PortletRepo(i0.inject(i1.HttpClient)); }, token: PortletRepo, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    PortletRepo.prototype.http;
}
/**
 * portlet 后端 -> 前端, 前端 -> 后端 转换
 * @param {?} serverPortlet
 * @return {?}
 */
function convertPortlet(serverPortlet) {
    serverPortlet.widget = convertWidget(serverPortlet.widget);
    if (serverPortlet.propValues)
        serverPortlet.propValues = JSON.parse(serverPortlet.propValues);
    return (/** @type {?} */ (Object.assign(new EppPortlet(null), serverPortlet)));
}
/**
 * @param {?} portlet
 * @return {?}
 */
function convertPortletInverse(portlet) {
    /** @type {?} */
    const serverPortlet = Object.assign({}, portlet);
    delete serverPortlet.widget;
    serverPortlet.propValues = (/** @type {?} */ (JSON.stringify(portlet.propValues)));
    return serverPortlet;
}
