// @farris/extend-file-upload/esm2015/lib/components/multi-download-info/multi-download-info.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/multi-download-info/multi-download-info.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ElementRef, ViewChild, HostBinding } from '@angular/core';
export class MultiDownloadInfoComponent {
    constructor() {
        this.cls = true;
        // 记载下载数据
        this._downloadInfos = [];
        // 默认名称
        this._downloadName = '';
        this.fileTypeCls = '';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set downloadInfos(value) {
        if (value && value.length > 0) {
            this._downloadInfos = value;
            if (!this._downloadName) {
                this.downloadName = this.getName(this._downloadInfos[0]['name']);
            }
        }
    }
    /**
     * @return {?}
     */
    get downloadInfos() {
        return this._downloadInfos;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set downloadName(value) {
        this._downloadName = value;
    }
    /**
     * @return {?}
     */
    get downloadName() {
        if (!this._downloadName && this._downloadInfos.length > 0) {
            this._downloadName = this.getName(this._downloadInfos[0]['name']);
        }
        return this._downloadName;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        //this.multiRenameRef.nativeElement.value = this.downloadName;
    }
    /**
     * 获取下载名称
     * @return {?}
     */
    getDownloadName() {
        return this.multiRenameRef.nativeElement.value;
    }
    /**
     * @private
     * @param {?} name
     * @return {?}
     */
    getName(name) {
        if (name.lastIndexOf('.') > 0) {
            return name.substring(0, name.lastIndexOf('.'));
        }
        return '';
    }
    /**
     * @param {?} filename
     * @return {?}
     */
    getFileTypeClassName(filename) {
    }
}
MultiDownloadInfoComponent.decorators = [
    { type: Component, args: [{
                selector: 'ffile-multi-download-info',
                template: "<ng-container *ngIf=\"downloadInfos.length>0\">\r\n  <div class=\"ffile-download-info\">\r\n    <div class=\"ffile-download-info--header\" style=\"align-items: center;\">\r\n      <span style=\"width: 22px;height: 22px;margin-right: 5px;\" class=\"ffilepreview--filetype-icon\" [ngClass]=\"fileTypeCls\"></span>\r\n      <span class=\"file-name\">\r\n        {{downloadInfos[0]['name']}}\r\n      </span>\r\n      <span>{{'uploadAndPreview.downloadCount' | upLocale: downloadInfos.length}}</span>\r\n    </div>\r\n    <div class=\"ffile-download-info--content\">\r\n      <span class=\"rename-label\">{{'uploadAndPreview.newName' | upLocale }}</span>\r\n      <div class=\"rename-container\"><input type=\"text\"\r\n        class=\"form-control\" placeholder=\"\" #multiRenameInput  [value]=\"downloadName\" /></div>\r\n      <span class=\"rename-suffix\">.zip</span>\r\n    </div>\r\n  </div>\r\n</ng-container>",
                styles: [".ffile-download-info--header{display:-webkit-box;display:flex;margin:0 0 16px}.ffile-download-info--header .file-name{overflow:hidden;text-overflow:ellipsis;white-space:nowrap;-webkit-box-flex:1;flex:1 1 0;padding-right:8px}.ffile-download-info--content{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center}.ffile-download-info--content .rename-container{-webkit-box-flex:1;flex:1 1 0}.ffile-download-info{padding:10px 24px;font-size:13px}.rename-label{color:#5a5e66;margin:0 10px 0 0}.rename-suffix{margin:0 0 0 8px}"]
            }] }
];
/** @nocollapse */
MultiDownloadInfoComponent.ctorParameters = () => [];
MultiDownloadInfoComponent.propDecorators = {
    cls: [{ type: HostBinding, args: ['class.ffile-multi-download-info',] }],
    multiRenameRef: [{ type: ViewChild, args: ['multiRenameInput',] }],
    downloadInfos: [{ type: Input }],
    downloadName: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    MultiDownloadInfoComponent.prototype.cls;
    /** @type {?} */
    MultiDownloadInfoComponent.prototype.multiRenameRef;
    /**
     * @type {?}
     * @private
     */
    MultiDownloadInfoComponent.prototype._downloadInfos;
    /**
     * @type {?}
     * @private
     */
    MultiDownloadInfoComponent.prototype._downloadName;
    /** @type {?} */
    MultiDownloadInfoComponent.prototype.fileTypeCls;
}
