// @gspwidget/widget-core/esm5/lib/instance.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * instance 指运行时，在某处运行的一个部件。
 * 此服务被动收集、输出实例id与实例引用信息，
 * 主动调用者为 widgetComponent 或 portletComponent
 */
var InstanceService = /** @class */ (function () {
    function InstanceService() {
        /**
         * 部件实例引用信息
         */
        this.instances = {};
    }
    /**
     * @param {?} insId
     * @return {?}
     */
    InstanceService.prototype.getInsByPInsId = /**
     * @param {?} insId
     * @return {?}
     */
    function (insId) {
        return this.instances[insId];
    };
    /**
     * @param {?} insId
     * @param {?} widgetComp
     * @param {?} type
     * @return {?}
     */
    InstanceService.prototype.addInsInfo = /**
     * @param {?} insId
     * @param {?} widgetComp
     * @param {?} type
     * @return {?}
     */
    function (insId, widgetComp, type) {
        if (!insId)
            console.warn('ins id 为空');
        if (!this.instances[insId])
            this.instances[insId] = {};
        this.instances[insId][type] = widgetComp;
    };
    InstanceService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    InstanceService.ctorParameters = function () { return []; };
    /** @nocollapse */ InstanceService.ngInjectableDef = i0.defineInjectable({ factory: function InstanceService_Factory() { return new InstanceService(); }, token: InstanceService, providedIn: "root" });
    return InstanceService;
}());
export { InstanceService };
if (false) {
    /**
     * 部件实例引用信息
     * @type {?}
     */
    InstanceService.prototype.instances;
}
