// @gspwidget/widget-core/esm5/lib/widget-core.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { UtilModule } from '@gspwidget/util';
import { WidgetDevkitModule } from '@gspwidget/widget-devkit';
import { TranslateModule } from '@ngx-translate/core';
import { NzAutocompleteModule } from 'ng-zorro-antd/auto-complete';
import { NzDatePickerModule } from 'ng-zorro-antd/date-picker';
import { NzModalModule } from 'ng-zorro-antd/modal';
import { AccordionModule } from 'ngx-bootstrap/accordion';
import { ColorPickerModule } from 'ngx-color-picker';
import { WidgetComponent } from './widget/widget.component';
import { TitleBarComponent } from './title-bar/title-bar.component';
import { FilterModule } from './filter/filter.module';
import { NzDropDownModule } from 'ng-zorro-antd/dropdown';
import { NzSelectModule } from 'ng-zorro-antd/select';
import { NzInputModule } from 'ng-zorro-antd/input';
// import { FilterComponent } from './filter/filter.component';
// import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
// import { GridDesignerComponent } from '../../../designer/src/app/components/grid-designer/grid-designer.component';
// import {GridsterModule as Gridster2} from '@widget/gridster';
// TODO: 引用方式
// import { WidgetDevkitModule } from '@gspwidget/widget-devkit';
var WidgetCoreModule = /** @class */ (function () {
    function WidgetCoreModule() {
    }
    WidgetCoreModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        WidgetComponent,
                        TitleBarComponent,
                    ],
                    entryComponents: [],
                    imports: [
                        CommonModule,
                        FormsModule,
                        HttpClientModule,
                        ReactiveFormsModule,
                        AccordionModule.forRoot(),
                        ColorPickerModule,
                        // Gridster2
                        // TranslateModule.forRoot(),
                        TranslateModule,
                        NzAutocompleteModule,
                        NzModalModule,
                        NzDatePickerModule,
                        // NzSelectModule,
                        // 保证元部件在这层能获取到这里面的服务。
                        // 不能仅注册在 root，因为服务的服务不保证在 root 上；
                        // 不能都注册在 widget 里，因为会每个部件一个服务实例；
                        // 注册在这保证元部件在此层找到同一个服务实例、并且服务的服务也在 module 里提供了。
                        WidgetDevkitModule,
                        UtilModule,
                        FilterModule,
                        NzDropDownModule,
                        NzSelectModule,
                        NzInputModule,
                    ],
                    exports: [
                        WidgetComponent,
                    ],
                    providers: []
                },] }
    ];
    return WidgetCoreModule;
}());
export { WidgetCoreModule };
