// @farris/ui-tabs/esm2015/lib/dropdown/dropdown.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NgbDropdown, NgbDropdownAnchor, NgbDropdownToggle, NgbDropdownMenu } from './dropdown';
import { NgbDropdownConfig } from './dropdown-config';
export { NgbDropdown, NgbDropdownToggle, NgbDropdownMenu } from './dropdown';
export { NgbDropdownConfig } from './dropdown-config';
import { FarrisCommonModule } from '@farris/ui-common';
/** @type {?} */
const NGB_DROPDOWN_DIRECTIVES = [
    NgbDropdown,
    NgbDropdownAnchor,
    NgbDropdownToggle,
    NgbDropdownMenu
];
export class NgbDropdownModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return { ngModule: NgbDropdownModule, providers: [NgbDropdownConfig] };
    }
}
NgbDropdownModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    FarrisCommonModule
                ],
                declarations: NGB_DROPDOWN_DIRECTIVES,
                exports: NGB_DROPDOWN_DIRECTIVES
            },] }
];
