// @gsp-wf/wf-process-editor/esm2015/lib/designer/wf-cell/OperationActivityCell.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { OperationActivity } from "@gsp-wf/wf-process-model";
import { UserActivityCell } from "./UserActivityCell";
import { FlowNodeState } from "@edp-pmf/bpmn-model";
import { WfConstants } from "../WfConstants";
/**
 * 人工操作活动节点
 */
export class OperationActivityCell extends UserActivityCell {
    /**
     * @param {?=} value
     * @param {?=} geometry
     * @param {?=} style
     */
    constructor(value, geometry, style) {
        super(value, geometry, style);
        this.clrTypeId = OperationActivityCell.CLR_TYPE_ID;
        this.name = "人工操作";
        this.nameLanguage = {
            "zh-CHS": "人工操作",
            "zh-CHT": "人工操作",
            "en": "Operation",
        };
        this.imgUrls = {
            [FlowNodeState.Default]: WfConstants.IMAGE_PATH + "/bizActivity.png",
            [FlowNodeState.Selected]: WfConstants.IMAGE_PATH + "/bizActivity_select.svg",
        };
    }
    /**
     * @param {?=} flowElt
     * @return {?}
     */
    createFlowElement(flowElt) {
        /** @type {?} */
        const flowNode = (/** @type {?} */ (super.createFlowElement(flowElt)));
        return flowNode;
    }
    /**
     * @return {?}
     */
    getFarrisPropConfig() {
        return this.getFarrisPropConfigByIsOperation(true);
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    updateProps(obj) {
        this.updatePropsByIsOperation(obj, true);
    }
}
OperationActivityCell.CLR_TYPE_ID = OperationActivity.Clr_Type_ID;
if (false) {
    /** @type {?} */
    OperationActivityCell.CLR_TYPE_ID;
    /** @type {?} */
    OperationActivityCell.prototype.clrTypeId;
    /** @type {?} */
    OperationActivityCell.prototype.name;
    /** @type {?} */
    OperationActivityCell.prototype.nameLanguage;
    /** @type {?} */
    OperationActivityCell.prototype.flowElement;
    /** @type {?} */
    OperationActivityCell.prototype.imgUrls;
}
