// @gsp-lcm/dbo-selectrt/esm2015/lib/dbo-selectrt/dbo-selectrt.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, TemplateRef } from '@angular/core';
import { DboHelpRtService } from '../dbo-help-rt.service';
export class DboSelectRTComponent {
    /**
     * @param {?} dboHelpService
     */
    constructor(dboHelpService) {
        this.dboHelpService = dboHelpService;
        this.url = '';
        /**
         * 帮助标题
         */
        this.title = '帮助';
        /**
         * 输出事件
         */
        this.selectedAfterOk = new EventEmitter();
        this.modalClosed = new EventEmitter();
        this.selectableSettings = {
            checkboxOnly: true,
            mode: 'single',
            enabled: true
        };
        /** @type {?} */
        const sessionId = this.sessionId;
        if (!sessionId) {
            alert('session信息不能为空');
        }
        else {
            this.dboHelpService.sessionId = sessionId;
        }
    }
    /**
     * 输入属性变化后
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.dboHelpService.getAllDatabaseObjectRT()
            .subscribe((/**
         * @param {?} result
         * @return {?}
         */
        result => {
            this.data = result;
            this.curSelectedKeys = [];
            if (this.initSelectedKeys) {
                this.initSelectedKeys.map((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => this.curSelectedKeys.push(item)));
            }
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.log('获取数据源失败！');
            console.log(error);
        }));
    }
    /**
     * 帮助确定按钮
     * @return {?}
     */
    btnOk() {
        // 根据选中行获取实体数据
        // 服务端获取数据
        /** @type {?} */
        const selectPosEntity = [];
        this.curSelectedKeys.forEach((/**
         * @param {?} dboID
         * @return {?}
         */
        dboID => {
            /** @type {?} */
            const selectDBO = this.data.find((/**
             * @param {?} item
             * @return {?}
             */
            item => item['id'] === dboID));
            selectPosEntity.push(selectDBO);
        }));
        if (!selectPosEntity) {
            alert('请选择数据库对象');
            return;
        }
        else {
            // 发射当前选中数据
            this.selectedAfterOk.next(selectPosEntity);
        }
        this.hideModal();
    }
    /**
     * @return {?}
     */
    hideModal() {
        this.modalClosed.emit();
    }
}
DboSelectRTComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-dbo-selectrt',
                template: "<div class=\"modal-body\">\r\n<div class=\"templateBlock\">\r\n  <kendo-grid [class]=\"'flex-auto'\" [selectable]=\"selectableSettings\" [kendoGridSelectBy]=\"'id'\" [selectedKeys]=\"curSelectedKeys\"\r\n    [kendoGridBinding]=\"data\">\r\n    <kendo-grid-checkbox-column width=\"10%\"></kendo-grid-checkbox-column>\r\n    <kendo-grid-column field=\"code\" title=\"\u7F16\u53F7\" width=\"40%\"></kendo-grid-column>\r\n    <kendo-grid-column field=\"name\" title=\"\u540D\u79F0\" width=\"45%\"></kendo-grid-column>\r\n  </kendo-grid>\r\n  </div>\r\n</div>\r\n<ng-template #btns>\r\n<button type=\"button\" class=\"btn btn-default btn-primary\" (click)=\"btnOk()\">\u786E\u5B9A</button>\r\n<button type=\"button\" class=\"btn btn-default\" (click)=\"hideModal()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
DboSelectRTComponent.ctorParameters = () => [
    { type: DboHelpRtService }
];
DboSelectRTComponent.propDecorators = {
    url: [{ type: Input }],
    title: [{ type: Input }],
    isShow: [{ type: Input }],
    initSelectedKeys: [{ type: Input }],
    sessionId: [{ type: Input }],
    selectedAfterOk: [{ type: Output }],
    modalClosed: [{ type: Output }],
    modalFooter: [{ type: ViewChild, args: ['modalFooter',] }],
    btns: [{ type: ViewChild, args: ['btns',] }]
};
if (false) {
    /** @type {?} */
    DboSelectRTComponent.prototype.url;
    /**
     * 帮助标题
     * @type {?}
     */
    DboSelectRTComponent.prototype.title;
    /**
     * 帮助打开状态
     * @type {?}
     */
    DboSelectRTComponent.prototype.isShow;
    /**
     * 选中列表
     * @type {?}
     */
    DboSelectRTComponent.prototype.initSelectedKeys;
    /** @type {?} */
    DboSelectRTComponent.prototype.sessionId;
    /**
     * 输出事件
     * @type {?}
     */
    DboSelectRTComponent.prototype.selectedAfterOk;
    /** @type {?} */
    DboSelectRTComponent.prototype.modalClosed;
    /** @type {?} */
    DboSelectRTComponent.prototype.modalFooter;
    /** @type {?} */
    DboSelectRTComponent.prototype.btns;
    /**
     * 当前界面选中的内码
     * @type {?}
     */
    DboSelectRTComponent.prototype.curSelectedKeys;
    /** @type {?} */
    DboSelectRTComponent.prototype.selectableSettings;
    /** @type {?} */
    DboSelectRTComponent.prototype.data;
    /**
     * @type {?}
     * @private
     */
    DboSelectRTComponent.prototype.dboHelpService;
}
