// @farris/extend-file-upload/esm5/lib/model/f-file-upload.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/model/f-file-upload.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function UploaderOptions() { }
if (false) {
    /** @type {?|undefined} */
    UploaderOptions.prototype.concurrency;
    /** @type {?|undefined} */
    UploaderOptions.prototype.allowedContentTypes;
    /** @type {?|undefined} */
    UploaderOptions.prototype.maxUploads;
    /** @type {?|undefined} */
    UploaderOptions.prototype.maxFileSize;
    /** @type {?|undefined} */
    UploaderOptions.prototype.uploadedCount;
}
/**
 * @record
 */
export function BlobFile() { }
if (false) {
    /** @type {?} */
    BlobFile.prototype.name;
}
/** @enum {number} */
var UploadStatus = {
    Queue: 0,
    Uploading: 1,
    Done: 2,
    Cancelled: 3,
    Remove: 4,
    Error: 5,
};
export { UploadStatus };
UploadStatus[UploadStatus.Queue] = 'Queue';
UploadStatus[UploadStatus.Uploading] = 'Uploading';
UploadStatus[UploadStatus.Done] = 'Done';
UploadStatus[UploadStatus.Cancelled] = 'Cancelled';
UploadStatus[UploadStatus.Remove] = 'Remove';
UploadStatus[UploadStatus.Error] = 'Error';
/**
 * @record
 */
export function UploadProgress() { }
if (false) {
    /** @type {?} */
    UploadProgress.prototype.status;
    /** @type {?|undefined} */
    UploadProgress.prototype.data;
}
/**
 * @record
 */
export function FUploadFileExtend() { }
if (false) {
    /** @type {?} */
    FUploadFileExtend.prototype.id;
    /** @type {?} */
    FUploadFileExtend.prototype.name;
    /** @type {?|undefined} */
    FUploadFileExtend.prototype.disabled;
    /** @type {?|undefined} */
    FUploadFileExtend.prototype.checked;
    /** @type {?|undefined} */
    FUploadFileExtend.prototype.size;
    /** @type {?|undefined} */
    FUploadFileExtend.prototype.createTime;
    /** @type {?|undefined} */
    FUploadFileExtend.prototype.type;
    /** @type {?|undefined} */
    FUploadFileExtend.prototype.extend;
    /** @type {?|undefined} */
    FUploadFileExtend.prototype.extendStatus;
    /** @type {?|undefined} */
    FUploadFileExtend.prototype.extendHeaders;
    /** @type {?|undefined} */
    FUploadFileExtend.prototype.showDownload;
    /** @type {?|undefined} */
    FUploadFileExtend.prototype.showComments;
    /** @type {?|undefined} */
    FUploadFileExtend.prototype.modeless;
}
/**
 * @record
 */
export function UploadFile() { }
if (false) {
    /** @type {?} */
    UploadFile.prototype.id;
    /** @type {?} */
    UploadFile.prototype.fileIndex;
    /** @type {?} */
    UploadFile.prototype.lastModifiedDate;
    /** @type {?} */
    UploadFile.prototype.name;
    /** @type {?} */
    UploadFile.prototype.size;
    /** @type {?} */
    UploadFile.prototype.type;
    /** @type {?} */
    UploadFile.prototype.form;
    /** @type {?} */
    UploadFile.prototype.progress;
    /** @type {?|undefined} */
    UploadFile.prototype.response;
    /** @type {?|undefined} */
    UploadFile.prototype.responseStatus;
    /** @type {?|undefined} */
    UploadFile.prototype.sub;
    /** @type {?|undefined} */
    UploadFile.prototype.nativeFile;
    /** @type {?|undefined} */
    UploadFile.prototype.responseHeaders;
}
/**
 * @record
 */
export function UploadOutput() { }
if (false) {
    /** @type {?} */
    UploadOutput.prototype.type;
    /** @type {?|undefined} */
    UploadOutput.prototype.file;
    /** @type {?|undefined} */
    UploadOutput.prototype.nativeFile;
    /** @type {?|undefined} */
    UploadOutput.prototype.message;
    /** @type {?|undefined} */
    UploadOutput.prototype.files;
}
/**
 * @record
 */
export function UploadConfig() { }
if (false) {
    /** @type {?|undefined} */
    UploadConfig.prototype.url;
    /** @type {?|undefined} */
    UploadConfig.prototype.method;
    /** @type {?|undefined} */
    UploadConfig.prototype.id;
    /** @type {?|undefined} */
    UploadConfig.prototype.fieldName;
    /** @type {?|undefined} */
    UploadConfig.prototype.fileIndex;
    /** @type {?|undefined} */
    UploadConfig.prototype.file;
    /** @type {?|undefined} */
    UploadConfig.prototype.data;
    /** @type {?|undefined} */
    UploadConfig.prototype.headers;
    /** @type {?|undefined} */
    UploadConfig.prototype.includeWebKitFormBoundary;
    /** @type {?|undefined} */
    UploadConfig.prototype.withCredentials;
    /** @type {?|undefined} */
    UploadConfig.prototype.timeout;
}
/**
 * @record
 */
export function UploadInput() { }
if (false) {
    /** @type {?} */
    UploadInput.prototype.type;
    /**
     * 分块上传时，每块大小默认为 1M
     * @type {?|undefined}
     */
    UploadInput.prototype.chunkSize;
}
/**
 * @abstract
 */
var /**
 * @abstract
 */
UploadServerService = /** @class */ (function () {
    function UploadServerService() {
    }
    return UploadServerService;
}());
/**
 * @abstract
 */
export { UploadServerService };
if (false) {
    /**
     * @abstract
     * @param {?} files
     * @param {?} event
     * @param {?} extendService
     * @return {?}
     */
    UploadServerService.prototype.upload = function (files, event, extendService) { };
    /**
     * @abstract
     * @param {?} files
     * @param {?} event
     * @param {?} extendService
     * @return {?}
     */
    UploadServerService.prototype.remove = function (files, event, extendService) { };
    /**
     * @abstract
     * @param {?} file
     * @param {?} event
     * @param {?} extendService
     * @return {?}
     */
    UploadServerService.prototype.multipartUpload = function (file, event, extendService) { };
}
/**
 * @record
 */
export function DocMetaProperty() { }
if (false) {
    /** @type {?|undefined} */
    DocMetaProperty.prototype.extProperties1;
    /** @type {?|undefined} */
    DocMetaProperty.prototype.extProperties2;
    /** @type {?|undefined} */
    DocMetaProperty.prototype.extProperties3;
    /** @type {?|undefined} */
    DocMetaProperty.prototype.extProperties4;
    /** @type {?|undefined} */
    DocMetaProperty.prototype.extProperties5;
}
