// @gsp-bef/gsp-cm-metadata/esm2015/lib/common-model/src/GSPCommonElement.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { GspCommonField, GSPElementDataType, GSPElementObjectType } from '../ref';
/**
 * 字段定义
 * @author wangjiegj<wangjiegj@inspur.com>
 */
/**
 * 字段定义
 */
export class GSPCommonElement extends GspCommonField {
    constructor() {
        super();
        /**
         * 字段对象类型(默认None)
         */
        this.ObjectType = GSPElementObjectType.None;
        // this.ID = Guid.newGuid();
        this.Length = 36;
        this.Precision = 0;
        this.MDataType = GSPElementDataType.String;
        this.ObjectType = GSPElementObjectType.None;
        this.IsMultiLanguage = false;
        this.IsRequire = false;
        this.IsRef = false;
    }
}
if (false) {
    /**
     * 编码规则配置
     * @type {?}
     */
    GSPCommonElement.prototype.BillCodeConfig;
    /**
     * 字段默认值类型
     * @type {?}
     */
    GSPCommonElement.prototype.DefaultVauleType;
    /**
     * 字段对象类型(默认None)
     * @type {?}
     */
    GSPCommonElement.prototype.ObjectType;
    /**
     * 对应的数据对象的列ID
     * @type {?}
     */
    GSPCommonElement.prototype.ColumnID;
    /**
     * 是否只读
     * @type {?}
     */
    GSPCommonElement.prototype.Readonly;
    /**
     * 是否自定义项
     * @type {?}
     */
    GSPCommonElement.prototype.IsCustomItem;
    /**
     * 包含的关联列表
     * @type {?}
     */
    GSPCommonElement.prototype.ChildAssociations;
}
