// @farris/mobile-command-services/esm5/lib/data-services/approve.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/data-services/approve.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { EMPTY } from 'rxjs';
import { NotifyService } from '../ui-services/index';
var ApproveService = /** @class */ (function () {
    function ApproveService(wfTaskHandlerService, viewModelContext) {
        this.viewModelContext = viewModelContext;
        this.wfTaskHandlerService = wfTaskHandlerService;
        this.notifyService = this.getService(NotifyService);
    }
    /**
     * 获取服务实例
     */
    /**
     * 获取服务实例
     * @template T
     * @param {?} token
     * @return {?}
     */
    ApproveService.prototype.getService = /**
     * 获取服务实例
     * @template T
     * @param {?} token
     * @return {?}
     */
    function (token) {
        /** @type {?} */
        var injector = this.viewModelContext.injector;
        return injector.get(token);
    };
    /**
     * 提交审批(带入口单据使用wf控件)
     * @param bizBillID 业务单据Id
     * @param bizDefKey 入口单据Id
     * @param options 上下文参数
     * @param variables 可选参数
     */
    /**
     * 提交审批(带入口单据使用wf控件)
     * @param {?} bizBillID 业务单据Id
     * @param {?} bizDefKey 入口单据Id
     * @param {?=} options 上下文参数
     * @param {?=} variables 可选参数
     * @return {?}
     */
    ApproveService.prototype.submitApproveWithBizDefKeyUseControl = /**
     * 提交审批(带入口单据使用wf控件)
     * @param {?} bizBillID 业务单据Id
     * @param {?} bizDefKey 入口单据Id
     * @param {?=} options 上下文参数
     * @param {?=} variables 可选参数
     * @return {?}
     */
    function (bizBillID, bizDefKey, options, variables) {
        if (options === void 0) { options = {}; }
        if (!bizBillID) {
            this.notifyService.error('业务单据Id不能为空');
            return EMPTY;
        }
        if (!bizDefKey) {
            this.notifyService.error('入口单据Id不能为空');
            return EMPTY;
        }
        if (!options || typeof options !== 'object') {
            options = {};
        }
        /** @type {?} */
        var payload = tslib_1.__assign({ dataId: bizBillID, bizDefKey: bizDefKey }, options);
        // 处理variables参数
        if (variables) {
            if (variables.startsWith('{') && variables.endsWith('}')) {
                try {
                    variables = JSON.parse(variables);
                }
                catch (_a) {
                    console.log('variables parse failed!');
                    variables = {};
                }
            }
            payload.variables = variables;
        }
        return this.wfTaskHandlerService && this.wfTaskHandlerService.startProcess(payload);
    };
    /**
     * 取消审批(支持主表、子表)
     * 迁移web端（已废弃）请调用新的取消提交方法
     * @param procInstId 流程实例Id
     */
    /**
     * 取消审批(支持主表、子表)
     * 迁移web端（已废弃）请调用新的取消提交方法
     * @param {?} procInstId 流程实例Id
     * @return {?}
     */
    ApproveService.prototype.cancelSubmit = /**
     * 取消审批(支持主表、子表)
     * 迁移web端（已废弃）请调用新的取消提交方法
     * @param {?} procInstId 流程实例Id
     * @return {?}
     */
    function (procInstId) {
        if (!procInstId) {
            this.notifyService.error('流程实例Id不能为空');
            return EMPTY;
        }
        return this.wfTaskHandlerService && this.wfTaskHandlerService.cancelSubmit({ dataId: null, bizDefKey: null, procInstId: procInstId });
    };
    /**
     * 提交审批
     * @param dataId 表单id
     * @param bizDefKey 入口单据id
     */
    /**
     * 提交审批
     * @param {?} dataId 表单id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    ApproveService.prototype.submitWithBizDefKey = /**
     * 提交审批
     * @param {?} dataId 表单id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    function (dataId, bizDefKey) {
        return this.wfTaskHandlerService && this.wfTaskHandlerService.submitWithBizDefKey(dataId, bizDefKey);
    };
    /**
     * 取消提交审批
     * @param dataId 表单id
     * @param bizDefKey 入口单据id
     */
    /**
     * 取消提交审批
     * @param {?} dataId 表单id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    ApproveService.prototype.cancelSubmitWithDataId = /**
     * 取消提交审批
     * @param {?} dataId 表单id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    function (dataId, bizDefKey) {
        return this.wfTaskHandlerService && this.wfTaskHandlerService.cancelSubmitWithDataId(dataId, bizDefKey);
    };
    /**
     * 批量提交审批
     * @param dataIds 表单ID集合
     * @param bizDefKey 入口单据id
     */
    /**
     * 批量提交审批
     * @param {?} dataIds 表单ID集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    ApproveService.prototype.batchSubmitWithBizDefKey = /**
     * 批量提交审批
     * @param {?} dataIds 表单ID集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    function (dataIds, bizDefKey) {
        return this.wfTaskHandlerService && this.wfTaskHandlerService.batchSubmitWithBizDefKey(dataIds, bizDefKey);
    };
    /**
     * 批量取消提交审批
     * @param dataIds 表单ID集合
     * @param bizDefKey 入口单据id
     */
    /**
     * 批量取消提交审批
     * @param {?} dataIds 表单ID集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    ApproveService.prototype.batchCancelSubmitWithDataId = /**
     * 批量取消提交审批
     * @param {?} dataIds 表单ID集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    function (dataIds, bizDefKey) {
        return this.wfTaskHandlerService && this.wfTaskHandlerService.batchCancelSubmitWithDataId(dataIds, bizDefKey);
    };
    /**
    * 子表提交审批
    * @param parentDataId 主表id
    * @param childDataId 子表id
    * @param bizDefKey 入口单据id
    */
    /**
     * 子表提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataId 子表id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    ApproveService.prototype.childSubmit = /**
     * 子表提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataId 子表id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    function (parentDataId, childDataId, bizDefKey) {
        return this.wfTaskHandlerService && this.wfTaskHandlerService.childSubmit(parentDataId, childDataId, bizDefKey);
    };
    /**
     * 子表取消提交审批
     * @param parentDataId 主表id
     * @param childDataId 子表id
     * @param bizDefKey 入口单据id
     */
    /**
     * 子表取消提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataId 子表id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    ApproveService.prototype.childCancelSubmit = /**
     * 子表取消提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataId 子表id
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    function (parentDataId, childDataId, bizDefKey) {
        return this.wfTaskHandlerService && this.wfTaskHandlerService.childCancelSubmit(parentDataId, childDataId, bizDefKey);
    };
    /**
     * 子表批量提交审批
     * @param parentDataId 主表id
     * @param childDataIds 子表id集合
     * @param bizDefKey 入口单据id
     */
    /**
     * 子表批量提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataIds 子表id集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    ApproveService.prototype.childBatchSubmit = /**
     * 子表批量提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataIds 子表id集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    function (parentDataId, childDataIds, bizDefKey) {
        return this.wfTaskHandlerService && this.wfTaskHandlerService.childBatchSubmit(parentDataId, childDataIds, bizDefKey);
    };
    /**
     * 子表批量取消提交审批
     * @param parentDataId 主表id
     * @param childDataIds 子表id集合
     * @param bizDefKey 入口单据id
     */
    /**
     * 子表批量取消提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataIds 子表id集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    ApproveService.prototype.childBatchCancelSubmit = /**
     * 子表批量取消提交审批
     * @param {?} parentDataId 主表id
     * @param {?} childDataIds 子表id集合
     * @param {?} bizDefKey 入口单据id
     * @return {?}
     */
    function (parentDataId, childDataIds, bizDefKey) {
        return this.wfTaskHandlerService && this.wfTaskHandlerService.childBatchCancelSubmit(parentDataId, childDataIds, bizDefKey);
    };
    return ApproveService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    ApproveService.prototype.wfTaskHandlerService;
    /**
     * @type {?}
     * @private
     */
    ApproveService.prototype.notifyService;
    /**
     * @type {?}
     * @protected
     */
    ApproveService.prototype.viewModelContext;
}
export { ApproveService };
