// @farris/mobile-command-services/esm5/lib/data-services/base-data.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/data-services/base-data.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EntityService } from '../entity-services/index';
import { LoadingService, ToastService, NotifyService, DialogService } from '../ui-services/index';
import { ExceptionService } from '../exception-services/index';
/**
 * 数据服务基类
 */
var /**
 * 数据服务基类
 */
BaseDataService = /** @class */ (function () {
    function BaseDataService(viewModelContext) {
        this.viewModelContext = viewModelContext;
        this.init();
    }
    /**
     * @private
     * @return {?}
     */
    BaseDataService.prototype.init = /**
     * @private
     * @return {?}
     */
    function () {
        this.initServices();
    };
    /**
     * @private
     * @return {?}
     */
    BaseDataService.prototype.initServices = /**
     * @private
     * @return {?}
     */
    function () {
        this.loadingService = this.getService(LoadingService);
        this.toastService = this.getService(ToastService);
        this.notifyService = this.getService(NotifyService);
        this.dialogService = this.getService(DialogService);
        this.exceptionService = this.getService(ExceptionService);
        this.entityService = this.getService(EntityService);
        this.befRepository = (/** @type {?} */ (this.viewModelContext.repository));
    };
    /**
     * 获取服务实例
     */
    /**
     * 获取服务实例
     * @template T
     * @param {?} token
     * @return {?}
     */
    BaseDataService.prototype.getService = /**
     * 获取服务实例
     * @template T
     * @param {?} token
     * @return {?}
     */
    function (token) {
        /** @type {?} */
        var injector = this.viewModelContext.injector;
        return injector.get(token);
    };
    return BaseDataService;
}());
if (false) {
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.loadingService;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.toastService;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.notifyService;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.dialogService;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.exceptionService;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.befRepository;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.entityService;
    /**
     * @type {?}
     * @protected
     */
    BaseDataService.prototype.viewModelContext;
}
export { BaseDataService };
